// @a0n/.ignored_gnosis/moonshine-tauri/src/index.tsx
import React, { useState, useEffect, useRef } from '@a0n/raect';
import { Archive, Trash2, Activity } from '@a0n/aeon-icons';
import {
  withPreferenceNamespace,
  type PreferenceNamespaceRecord,
} from '@a0n/aeon-preferences';
import { invoke } from '@tauri-apps/api/core';
import { listen } from '@tauri-apps/api/event';
import './index.css';

// Forward WebView console errors/warnings to the host stderr (the WebView's
// own console is invisible from a terminal launch). Install once at module
// load so renderer crashes during early mount are still captured.
(() => {
  const fwd = (level: string, args: unknown[]) => {
    try {
      const msg = args
        .map((a) => {
          if (a instanceof Error) return `${a.message}\n${a.stack ?? ''}`;
          if (typeof a === 'string') return a;
          try {
            return JSON.stringify(a);
          } catch {
            return String(a);
          }
        })
        .join(' ');
      invoke('js_log', { level, msg }).catch(() => {});
    } catch {
      /* never let logging break the app */
    }
  };
  const origErr = console.error.bind(console);
  const origWarn = console.warn.bind(console);
  console.error = (...args: unknown[]) => {
    fwd('error', args);
    origErr(...args);
  };
  console.warn = (...args: unknown[]) => {
    fwd('warn', args);
    origWarn(...args);
  };
  window.addEventListener('error', (e) => {
    fwd('uncaught', [e.message, `${e.filename}:${e.lineno}:${e.colno}`, e.error]);
  });
  window.addEventListener('unhandledrejection', (e) => {
    fwd('rejection', [e.reason]);
  });
})();

interface ProcessInfo {
  id: string;
  command: string;
  status: string;
  created_at: string;
  exit_code: number | null;
  output: string;
  duration_ms: number | null;
}

interface PtySessionInfo {
  id: string;
  command: string;
  executedCommand: string;
  created_at: string;
  running: boolean;
}

interface SystemInfo {
  moonshine_bin: string;
  guard_bin: string;
  guard_home: string;
  identity_present: boolean;
  identity_path: string;
  grants_count: number;
  repl_active: boolean;
  ready: boolean;
  problems: string[];
}

interface ModelSnapshot {
  id: string;
  label: string;
  url: string;
  cache_path: string;
  present: boolean;
  size_bytes: number | null;
  active: boolean;
}

interface ModelStatus {
  active_model_id: string;
  cache_dir: string;
  models: ModelSnapshot[];
}

interface SkymeshStatus {
  enabled: boolean;
  running: boolean;
  pid: number | null;
  node: string | null;
  model: string | null;
  last_cache_key: string | null;
  filled_cache_keys: string[];
  last_error: string | null;
  updated_at: string | null;
}

interface MobileBridgeStatus {
  enabled: boolean;
  url: string;
  token_configured: boolean;
  config_path: string;
  last_error: string | null;
}

interface Admission {
  name: string;
  caps: string[];
  expires_at: number | null;
  expired: boolean;
  revoked: boolean;
}

const ALL_CAPS = ['fs', 'net', 'exec'] as const;
const STARTER_COMMAND = 'what is the capital of France?';

const AGENTIC_MODALITIES = [
  {
    id: 'todos',
    label: 'Todos',
    title: 'open Fractal Todo agentic mode',
    command:
      'agent exec --permission-mode ask --max-turns 2 "Use the MCP catalog to find Fractal Todo tools, summarize the task surface, then wait for my next instruction."',
  },
  {
    id: 'notes',
    label: 'Notes',
    title: 'open Fractal Note agentic mode',
    command:
      'agent exec --permission-mode ask --max-turns 2 "Use the MCP catalog to find Fractal Note tools, summarize the note surface, then wait for my next instruction."',
  },
  {
    id: 'memories',
    label: 'Memories',
    title: 'open Fractal Memory agentic mode',
    command:
      'agent exec --permission-mode ask --max-turns 2 "Use the MCP catalog to find Fractal Memory tools, summarize the memory surface, then wait for my next instruction."',
  },
] as const;

function agenticModalityCommand(input: string): string | null {
  const alias = input.trim().toLowerCase();
  const match = AGENTIC_MODALITIES.find(
    (modality) => modality.id === alias || modality.label.toLowerCase() === alias
  );
  return match?.command ?? null;
}

// Command history persists across restarts via localStorage (Tauri keeps the
// WebView's localStorage on disk). Capped so it can't grow unbounded.
const HISTORY_KEY = 'moonshine.cmd.history';
const HISTORY_CAP = 500;
const AGENT_GEN_TOKENS_KEY = 'moonshine.agentGenTokens';
const DEFAULT_AGENT_GEN_TOKENS = 256;
const MIN_AGENT_GEN_TOKENS = 1;
const MAX_AGENT_GEN_TOKENS = 4096;
const ACCENT_COLOR_KEY = 'moonshine.accentColor';
const DEFAULT_ACCENT_COLOR = '#00ff9f';
const ACCENT_SWATCHES = ['#00ff9f', '#00d9ff', '#d4af37', '#ff3860', '#caa7ff'];
const VIM_KEY = 'moonshine.vim';
const ZOOM_KEY = 'moonshine.zoom';
const XTERM_KEY = 'moonshine.xterm';
const MOONSHINE_SETTINGS_NAMESPACE = 'app://moonshine.forkjoin.ai/settings';

interface MoonshineSettings {
  accentColor: string;
  agentGenTokens: number;
  vimEnabled: boolean;
  zoom: number;
  xtermEnabled: boolean;
}

const DEFAULT_MOONSHINE_SETTINGS: MoonshineSettings = {
  accentColor: DEFAULT_ACCENT_COLOR,
  agentGenTokens: DEFAULT_AGENT_GEN_TOKENS,
  vimEnabled: false,
  zoom: 1,
  xtermEnabled: true,
};

function normalizeAccentColor(value: string): string {
  const clean = value.trim();
  return /^#[0-9a-fA-F]{6}$/.test(clean) ? clean.toLowerCase() : DEFAULT_ACCENT_COLOR;
}

function loadAccentColor(): string {
  try {
    const raw = localStorage.getItem(ACCENT_COLOR_KEY);
    return raw ? normalizeAccentColor(raw) : DEFAULT_ACCENT_COLOR;
  } catch {
    return DEFAULT_ACCENT_COLOR;
  }
}

function saveAccentColor(value: string): string {
  const normalized = normalizeAccentColor(value);
  try {
    localStorage.setItem(ACCENT_COLOR_KEY, normalized);
  } catch {
    /* localStorage unavailable — keep the in-memory setting */
  }
  return normalized;
}

function normalizeAgentGenTokens(value: number): number {
  if (!Number.isFinite(value)) return DEFAULT_AGENT_GEN_TOKENS;
  return Math.min(
    MAX_AGENT_GEN_TOKENS,
    Math.max(MIN_AGENT_GEN_TOKENS, Math.round(value))
  );
}

function loadAgentGenTokens(): number {
  try {
    const raw = localStorage.getItem(AGENT_GEN_TOKENS_KEY);
    return raw ? normalizeAgentGenTokens(Number(raw)) : DEFAULT_AGENT_GEN_TOKENS;
  } catch {
    return DEFAULT_AGENT_GEN_TOKENS;
  }
}

function saveAgentGenTokens(value: number): number {
  const normalized = normalizeAgentGenTokens(value);
  try {
    localStorage.setItem(AGENT_GEN_TOKENS_KEY, String(normalized));
  } catch {
    /* localStorage unavailable — keep the in-memory setting */
  }
  return normalized;
}

function normalizeZoom(value: number): number {
  if (!Number.isFinite(value)) return DEFAULT_MOONSHINE_SETTINGS.zoom;
  return Math.min(2, Math.max(0.6, Math.round(value * 100) / 100));
}

function loadBooleanSetting(key: string, fallback: boolean): boolean {
  try {
    const raw = localStorage.getItem(key);
    if (raw === '1') return true;
    if (raw === '0') return false;
    return fallback;
  } catch {
    return fallback;
  }
}

function loadMoonshineSettings(): MoonshineSettings {
  let zoom = DEFAULT_MOONSHINE_SETTINGS.zoom;
  try {
    zoom = normalizeZoom(Number(localStorage.getItem(ZOOM_KEY) || zoom));
  } catch {
    zoom = DEFAULT_MOONSHINE_SETTINGS.zoom;
  }

  return {
    accentColor: loadAccentColor(),
    agentGenTokens: loadAgentGenTokens(),
    vimEnabled: loadBooleanSetting(VIM_KEY, DEFAULT_MOONSHINE_SETTINGS.vimEnabled),
    zoom,
    xtermEnabled: loadBooleanSetting(XTERM_KEY, DEFAULT_MOONSHINE_SETTINGS.xtermEnabled),
  };
}

function normalizeMoonshineSettings(
  value: unknown,
  fallback: MoonshineSettings = DEFAULT_MOONSHINE_SETTINGS
): MoonshineSettings {
  const source =
    typeof value === 'object' && value !== null && !Array.isArray(value)
      ? (value as Partial<MoonshineSettings>)
      : {};

  return {
    accentColor:
      typeof source.accentColor === 'string'
        ? normalizeAccentColor(source.accentColor)
        : fallback.accentColor,
    agentGenTokens:
      typeof source.agentGenTokens === 'number'
        ? normalizeAgentGenTokens(source.agentGenTokens)
        : fallback.agentGenTokens,
    vimEnabled:
      typeof source.vimEnabled === 'boolean'
        ? source.vimEnabled
        : fallback.vimEnabled,
    zoom:
      typeof source.zoom === 'number'
        ? normalizeZoom(source.zoom)
        : fallback.zoom,
    xtermEnabled:
      typeof source.xtermEnabled === 'boolean'
        ? source.xtermEnabled
        : fallback.xtermEnabled,
  };
}

function saveMoonshineSettings(settings: MoonshineSettings): MoonshineSettings {
  const normalized = normalizeMoonshineSettings(settings);
  saveAccentColor(normalized.accentColor);
  saveAgentGenTokens(normalized.agentGenTokens);
  try {
    localStorage.setItem(VIM_KEY, normalized.vimEnabled ? '1' : '0');
    localStorage.setItem(ZOOM_KEY, String(normalized.zoom));
    localStorage.setItem(XTERM_KEY, normalized.xtermEnabled ? '1' : '0');
  } catch {
    /* localStorage unavailable — keep the in-memory settings */
  }
  return normalized;
}

function moonshineSettingsNamespaceRecord(
  settings: MoonshineSettings
): PreferenceNamespaceRecord {
  return {
    accentColor: settings.accentColor,
    agentGenTokens: settings.agentGenTokens,
    vimEnabled: settings.vimEnabled,
    zoom: settings.zoom,
    xtermEnabled: settings.xtermEnabled,
  };
}

function loadHistory(): string[] {
  try {
    const raw = localStorage.getItem(HISTORY_KEY);
    const arr = raw ? JSON.parse(raw) : [];
    return Array.isArray(arr) ? arr.slice(-HISTORY_CAP) : [];
  } catch {
    return [];
  }
}
function saveHistory(h: string[]): string[] {
  const capped = h.slice(-HISTORY_CAP);
  try {
    localStorage.setItem(HISTORY_KEY, JSON.stringify(capped));
  } catch {
    /* quota or unavailable — history stays in-memory */
  }
  return capped;
}

// Programs that read their own stdin / want a TTY. The persistent REPL has no
// TTY and doesn't pipe stdin to children, so these starve and time out rather
// than going interactive. We warn before running and (for known tools) suggest
// the non-interactive arg form. Real fix is a per-command PTY path.
const INTERACTIVE_PROGRAMS = new Set([
  'claude', 'codex', 'gemini', 'vim', 'vi', 'nvim', 'nano', 'emacs',
  'less', 'more', 'top', 'htop', 'python', 'python3', 'node', 'irb',
  'ssh', 'tmux', 'fzf', 'man', 'zsh', 'bash', 'sh',
]);

/// True if the command's leading program is known-interactive and should run
/// in a PTY pane rather than the non-interactive REPL.
function isInteractiveCommand(command: string): boolean {
  const trimmed = command.trim();
  const first = trimmed.split(/\s+/)[0]?.split('/').pop() ?? '';
  if (first === 'agent' || trimmed === 'agentic') {
    return true;
  }
  return INTERACTIVE_PROGRAMS.has(first);
}

function ptyExecutionCommand(command: string): string {
  const trimmed = command.trim();
  if (trimmed === 'agentic') {
    return 'moonshine -c agentic';
  }
  // `agent`/`agentic` SUBCOMMANDS (exec, spawn, …) must go through moonshine's
  // REPL handler, which supports --permission-mode / --max-turns. Run bare in the
  // shell, `agent` resolves to the Cursor agent CLI (~/.local/bin/agent, a
  // commander.js shim) which rejects those flags with "unknown option
  // '--permission-mode'". Wrap the whole line into `moonshine -c '<line>'`.
  const firstWord = trimmed.split(/\s+/)[0];
  if ((firstWord === 'agent' || firstWord === 'agentic') && /\s/.test(trimmed)) {
    const escaped = trimmed.replace(/'/g, "'\\''");
    return `moonshine -c '${escaped}'`;
  }
  return command;
}

/// A VM boot target: `boot-iso <iso>`, `vm [boot] <iso>`, or a bare `.iso`/`.img`
/// path. Returns the disk-image path to boot, or null if this isn't a VM command.
/// A booted VM becomes its own persistent terminal tab (it keeps running in the
/// background) via `boot-iso` → the sovereign runner → Virtualization.framework.
function vmTarget(command: string): string | null {
  const t = command.trim();
  if (t.startsWith('boot-iso ')) return t.slice('boot-iso '.length).trim();
  if (t.startsWith('vm boot ')) return t.slice('vm boot '.length).trim();
  if (t.startsWith('vm ')) return t.slice('vm '.length).trim();
  if (/\s/.test(t)) return null;
  // A sovereign address, an ISO URL, or a local .iso/.img path — all bootable.
  if (t.startsWith('aeon://')) return t;
  if (/^https?:\/\/.+\.(iso|img)(\?.*)?$/i.test(t)) return t;
  if (/\.(iso|img)$/i.test(t)) return t;
  return null;
}

/// Where `aeon://iso/<id>` resolves: the edgework R2 ISO route (its own CDN).
const ISO_CDN = 'https://edge.forkjoin.ai/api/v1/r2/edgework-cache/isos';

/// Resolve a boot target to a fetchable URL or local path. An `aeon://` address
/// maps to its R2/CDN ISO URL by id (`aeon://iso/<id>` or
/// `aeon://rhizome/<world>/n/<id>`); URLs and local paths pass through.
function resolveBootUrl(target: string): string {
  const t = target.trim().replace(/^['"]|['"]$/g, '');
  if (t.startsWith('aeon://')) {
    const id = t.slice('aeon://'.length).split('/').filter(Boolean).pop() || t;
    const name = /\.(iso|img)$/i.test(id) ? id : `${id}.iso`;
    return `${ISO_CDN}/${name}`;
  }
  return t;
}

/// Single-quote a path for a POSIX shell.
function shQuote(p: string): string {
  return `'${p.replace(/'/g, "'\\''")}'`;
}

/// Last path component (or aeon id), for a VM tab label.
function baseName(p: string): string {
  const cleaned = p.replace(/['"]/g, '').replace(/\?.*$/, '').trim();
  return cleaned.split('/').filter(Boolean).pop() || cleaned || 'vm';
}

/// The command input doubles as a URL bar: a single bare token that looks like
/// a URL (http(s)://, aeon://, or a bare domain like example.com) is browsed
/// Longest common prefix of a list of strings (for Tab completion: extend the
/// token only as far as all candidates agree, never guessing past that).
function longestCommonPrefix(strs: string[]): string {
  if (strs.length === 0) return '';
  let prefix = strs[0];
  for (let i = 1; i < strs.length; i++) {
    while (prefix && !strs[i].startsWith(prefix)) prefix = prefix.slice(0, -1);
    if (!prefix) break;
  }
  return prefix;
}

/// rather than run as a shell command. Returns the normalized URL, or null if
/// the input isn't a standalone URL. Multi-token input (e.g. `echo example.com`)
/// is never treated as a URL — only a lone address.
function asBrowsableUrl(command: string): string | null {
  const t = command.trim();
  if (/\s/.test(t)) return null; // single token only
  if (/^https?:\/\//i.test(t) || /^aeon:\/\//i.test(t)) return t;
  // Bare domain: label(.label)+ optional /path, with a known-ish TLD shape.
  // Require a dot and no leading scheme-y colon, and a path or 2+ labels.
  if (/^[a-z0-9-]+(\.[a-z0-9-]+)+(\/[^\s]*)?$/i.test(t) && !t.includes('://')) {
    return `https://${t}`;
  }
  return null;
}

function formatCwdLabel(cwd: string): string {
  const clean = cwd.trim().replace(/\/+$/, '') || '/';
  if (!clean || clean === '/' || /^\/+$/.test(clean)) return '';
  const label = clean.replace(/^\/Users\/[^/]+/, '~');
  return label === '/' ? '' : label;
}

/// Turn http(s):// URLs in rendered page text into clickable links, so the
/// browser is navigable: clicking calls onNav(url). Non-URL text is preserved
/// verbatim. Keeps it cheap — a single split on a URL regex.
const URL_IN_TEXT = /(https?:\/\/[^\s<>"')\]]+)/g;
function linkify(text: string, onNav: (url: string) => void): React.ReactNode[] {
  // split() with a capturing group keeps the URLs as odd-indexed parts — no
  // stateful .test() (the /g regex's lastIndex would otherwise drift).
  const parts = text.split(URL_IN_TEXT);
  return parts.map((part, i) =>
    i % 2 === 1 ? (
      <span
        key={i}
        className="page-link"
        onClick={() => onNav(part)}
        title={`open ${part}`}
      >
        {part}
      </span>
    ) : (
      part
    )
  );
}

/// Render extracted page text with lightweight markdown-ish structure so code
/// and headings stand out (React-style inline code blocks). Block-level:
/// fenced ```code```, indented code (4+ spaces), and ALL-CAPS-ish/short lines
/// stay plain. Inline: `backtick code` becomes monospace chips; URLs stay
/// clickable via linkify. A pragmatic first pass — the polyglot HTML→AST path
/// (task #17) is the structural follow-on.
function renderRich(text: string, onNav: (url: string) => void): React.ReactNode[] {
  const lines = text.split('\n');
  const out: React.ReactNode[] = [];
  let i = 0;
  let key = 0;
  while (i < lines.length) {
    const line = lines[i];
    // Fenced code block ```
    if (line.trim().startsWith('```')) {
      const buf: string[] = [];
      i++;
      while (i < lines.length && !lines[i].trim().startsWith('```')) {
        buf.push(lines[i]);
        i++;
      }
      i++; // skip closing fence
      out.push(
        <code key={key++} className="md-codeblock">
          {buf.join('\n')}
        </code>
      );
      continue;
    }
    // Indented code block (4+ spaces or a tab), grouped.
    if (/^( {4,}|\t)/.test(line) && line.trim()) {
      const buf: string[] = [];
      while (i < lines.length && (/^( {4,}|\t)/.test(lines[i]) || lines[i].trim() === '')) {
        buf.push(lines[i].replace(/^( {4}|\t)/, ''));
        i++;
      }
      out.push(
        <code key={key++} className="md-codeblock">
          {buf.join('\n').replace(/\n+$/, '')}
        </code>
      );
      continue;
    }
    // Regular line — render inline `code` + links, then a newline.
    out.push(<span key={key++}>{renderInline(line, onNav)}</span>);
    out.push('\n');
    i++;
  }
  return out;
}

/// Inline markdown for one line: `code` spans + clickable URLs.
function renderInline(line: string, onNav: (url: string) => void): React.ReactNode[] {
  // Split on backtick code spans first; linkify the non-code segments.
  const segs = line.split(/(`[^`]+`)/g);
  const out: React.ReactNode[] = [];
  segs.forEach((seg, i) => {
    if (seg.startsWith('`') && seg.endsWith('`') && seg.length > 1) {
      out.push(
        <code key={`c${i}`} className="md-inline-code">
          {seg.slice(1, -1)}
        </code>
      );
    } else {
      out.push(<span key={`t${i}`}>{linkify(seg, onNav)}</span>);
    }
  });
  return out;
}

// ---- fractal-iam (card.yoga) DID auth -------------------------------------
const FRACTAL_IAM_URL = 'https://card.yoga';

interface AuthIdentity {
  provider: string;
  email?: string;
  name?: string;
  picture?: string;
  username?: string;
}
interface AuthSession {
  rootDid: string;
  bootstrapToken: string;
  expiresAtMs: number;
  identity: AuthIdentity;
}

const AUTH_STORAGE_KEY = 'moonshine.auth.session';

function loadStoredSession(): AuthSession | null {
  try {
    const raw = localStorage.getItem(AUTH_STORAGE_KEY);
    if (!raw) return null;
    const s = JSON.parse(raw) as AuthSession;
    if (s.expiresAtMs && s.expiresAtMs < Date.now()) return null;
    return s;
  } catch {
    return null;
  }
}

async function checkpointMoonshineSettings(
  session: AuthSession | null,
  settings: MoonshineSettings
): Promise<void> {
  if (!session) {
    return;
  }

  const preferences = withPreferenceNamespace(
    null,
    MOONSHINE_SETTINGS_NAMESPACE,
    moonshineSettingsNamespaceRecord(settings)
  );

  await fetch(`${FRACTAL_IAM_URL}/iam/preferences/checkpoint`, {
    method: 'POST',
    headers: {
      'content-type': 'application/json',
      authorization: `Bearer ${session.bootstrapToken}`,
    },
    body: JSON.stringify({
      rootDid: session.rootDid,
      preferences,
    }),
  }).catch(() => {
    /* remote preference checkpoint is best-effort; local settings already changed */
  });
}

/// Run the card.yoga browser sign-in: start a session, open the provider's
/// sign-in URL in the system browser, then poll until the DID + token land.
async function runCardYogaSignin(provider: string): Promise<AuthSession> {
  const startRes = await fetch(`${FRACTAL_IAM_URL}/auth/browser/start`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      provider,
      appId: 'moonshine-tauri',
      mode: 'popup',
      returnOrigin: window.location.origin,
    }),
  });
  if (!startRes.ok) throw new Error(`auth start failed: ${startRes.status}`);
  const start = await startRes.json();
  if (!start.ok) throw new Error(start.error || 'auth start rejected');

  // Open the provider sign-in in the system browser via the backend (the
  // plugin-shell JS package isn't installed; the Rust open_url command is).
  await invoke('open_url', { url: start.signInUrl });

  // Poll for completion (up to 2 min).
  const deadline = Date.now() + 120_000;
  while (Date.now() < deadline) {
    await new Promise((r) => setTimeout(r, 1500));
    const st = await fetch(start.statusUrl).then((r) => r.json()).catch(() => null);
    if (!st) continue;
    if (st.status === 'complete' && st.auth) {
      const session = st.auth as AuthSession;
      localStorage.setItem(AUTH_STORAGE_KEY, JSON.stringify(session));
      return session;
    }
    if (st.status === 'failed') throw new Error(st.error || 'sign-in failed');
  }
  throw new Error('sign-in timed out');
}

const POLL_MS = 250;

/// Strip ANSI escape sequences (color/style codes) so moonshine's own colored
/// status lines render as clean text instead of raw `[90m...[0m` noise.
// eslint-disable-next-line no-control-regex
const ANSI_RE = /\x1b\[[0-9;]*m/g;
/**
 * Renders the strip Ansi view.
 */
export function stripAnsi(s: string): string {
  return s.replace(ANSI_RE, '');
}

// ---- ANSI SGR color rendering --------------------------------------------
// Real terminal output (claude, git, ls --color, compiler errors) is colored
// via SGR escapes. Rather than strip them, we parse the codes into styled
// spans so the terminal looks like a terminal.

// The 16 base colors (normal + bright), tuned to read well on pure black.
const ANSI_16 = [
  '#1a1a1a', '#ff5f56', '#5af78e', '#f3f99d', '#57c7ff', '#ff6ac1', '#9aedfe', '#cfcfcf',
  '#686868', '#ff5f56', '#5af78e', '#f3f99d', '#57c7ff', '#ff6ac1', '#9aedfe', '#ffffff',
];

/// Resolve a 256-color cube/grayscale index to a hex string.
function ansi256(n: number): string {
  if (n < 16) return ANSI_16[n];
  if (n >= 232) {
    const v = 8 + (n - 232) * 10;
    return `rgb(${v},${v},${v})`;
  }
  const c = n - 16;
  const r = Math.floor(c / 36),
    g = Math.floor((c % 36) / 6),
    b = c % 6;
  const ch = (x: number) => (x === 0 ? 0 : 55 + x * 40);
  return `rgb(${ch(r)},${ch(g)},${ch(b)})`;
}

interface SgrStyle {
  color?: string;
  background?: string;
  fontWeight?: string;
  fontStyle?: string;
  textDecoration?: string;
  opacity?: number;
}

/// Apply one SGR parameter sequence (the numbers between ESC[ and m) to a style.
function applySgr(style: SgrStyle, params: number[]): SgrStyle {
  const s = { ...style };
  for (let i = 0; i < params.length; i++) {
    const p = params[i];
    if (p === 0) {
      for (const k of Object.keys(s)) delete (s as Record<string, unknown>)[k];
    } else if (p === 1) s.fontWeight = 'bold';
    else if (p === 2) s.opacity = 0.6;
    else if (p === 3) s.fontStyle = 'italic';
    else if (p === 4) s.textDecoration = 'underline';
    else if (p === 22) {
      delete s.fontWeight;
      delete s.opacity;
    } else if (p === 23) delete s.fontStyle;
    else if (p === 24) delete s.textDecoration;
    else if (p >= 30 && p <= 37) s.color = ANSI_16[p - 30];
    else if (p >= 90 && p <= 97) s.color = ANSI_16[p - 90 + 8];
    else if (p >= 40 && p <= 47) s.background = ANSI_16[p - 40];
    else if (p >= 100 && p <= 107) s.background = ANSI_16[p - 100 + 8];
    else if (p === 39) delete s.color;
    else if (p === 49) delete s.background;
    else if (p === 38 || p === 48) {
      // Extended color: 38;5;n (256) or 38;2;r;g;b (truecolor).
      const target = p === 38 ? 'color' : 'background';
      if (params[i + 1] === 5) {
        s[target] = ansi256(params[i + 2] ?? 7);
        i += 2;
      } else if (params[i + 1] === 2) {
        s[target] = `rgb(${params[i + 2] ?? 0},${params[i + 3] ?? 0},${params[i + 4] ?? 0})`;
        i += 4;
      }
    }
  }
  return s;
}

/// Parse a string with ANSI SGR escapes into an array of styled React spans.
/// SGR sequences become styled spans; all other escapes (full CSI grammar
/// incl. private `?`/`>`/`<` prefixes and intermediates, OSC, two-char escapes
/// like ESC 7/8/=/>, charset selects) are consumed so they don't leak as
/// mojibake. This is scrollback-grade — it does NOT honor cursor addressing,
/// so full-screen TUIs (claude Code, vim) still need xterm.js (task #15).
/**
 * Renders the render Ansi view.
 */
export function renderAnsi(
  input: string,
  onNav?: (url: string) => void,
  find?: { query: string; current: number }
): React.ReactNode[] {
  const out: React.ReactNode[] = [];
  let style: SgrStyle = {};
  let last = 0;
  let key = 0;
  // Running index of find matches across the whole output, so the Nth match can
  // be marked "current" for scroll-into-view + cur/total display.
  let matchIdx = 0;
  const findQ = find && find.query ? find.query : '';
  const highlight = (text: string): React.ReactNode => {
    const lower = text.toLowerCase();
    const q = findQ.toLowerCase();
    const parts: React.ReactNode[] = [];
    let i = 0;
    let start = 0;
    let k = 0;
    for (;;) {
      const at = lower.indexOf(q, i);
      if (at === -1) {
        parts.push(text.slice(start));
        break;
      }
      if (at > start) parts.push(text.slice(start, at));
      const idx = matchIdx++;
      parts.push(
        <mark key={`m${k++}`} className={`find-hit${idx === find!.current ? ' current' : ''}`}>
          {text.slice(at, at + findQ.length)}
        </mark>
      );
      i = at + findQ.length;
      start = i;
    }
    return <React.Fragment key={key++}>{parts}</React.Fragment>;
  };
  // Order matters: match SGR first (capture params), then OSC, then any other
  // CSI (private-prefix + params + intermediates + final byte), then two-char
  // escapes. Final alternative strips stray C0 control chars except \n and \t.
  // eslint-disable-next-line no-control-regex
  const re =
    /\x1b\[([0-9;]*)m|\x1b\][\s\S]*?(?:\x07|\x1b\\)|\x1b\[[0-9;?<>=]*[ -/]*[@-~]|\x1b[ -/]*[0-9A-Za-z=><]|[\x00-\x08\x0b-\x1f\x7f]/g;
  let m: RegExpExecArray | null;
  const push = (text: string) => {
    if (!text) return;
    // Find-highlight takes precedence (transient mode); otherwise linkify URLs
    // when a navigator is supplied. Both wrap in a keyed fragment so per-call
    // child keys don't collide.
    const content: React.ReactNode = findQ
      ? highlight(text)
      : onNav ? (
        <React.Fragment key={key++}>{linkify(text, onNav)}</React.Fragment>
      ) : (
        text
      );
    if (Object.keys(style).length === 0) out.push(content);
    else out.push(
      <span key={key++} style={style as React.CSSProperties}>
        {content}
      </span>
    );
  };
  while ((m = re.exec(input)) !== null) {
    push(input.slice(last, m.index));
    last = m.index + m[0].length;
    if (m[1] !== undefined) {
      // SGR sequence — update style. Empty params ([m) means reset.
      const params = m[1] === '' ? [0] : m[1].split(';').map((n) => parseInt(n, 10) || 0);
      style = applySgr(style, params);
    }
    // Everything else (cursor moves, OSC, mode sets, control chars) is dropped.
  }
  push(input.slice(last));
  return out;
}

/// Astrolabe loading indicator — concentric golden rings spinning at different
/// rates with a breathing pulse, distilled from apps/astrolabe (whose real
/// visual is heavy Three.js). Pure SVG + CSS keyframes, zero deps.
function AstrolabeSpinner({ label }: { label?: string }) {
  return (
    <div className="astrolabe-loading">
      <svg viewBox="0 0 100 100" className="astrolabe-spinner" aria-label="loading">
        <g className="ar ar-1">
          <circle cx="50" cy="50" r="40" fill="none" stroke="#d4af37" strokeWidth="2" opacity="0.85" strokeDasharray="180 70" />
        </g>
        <g className="ar ar-2">
          <circle cx="50" cy="50" r="30" fill="none" stroke="#caa84b" strokeWidth="1.8" opacity="0.7" strokeDasharray="120 70" />
        </g>
        <g className="ar ar-3">
          <circle cx="50" cy="50" r="20" fill="none" stroke="#9aedfe" strokeWidth="1.5" opacity="0.6" strokeDasharray="70 50" />
        </g>
        <g className="ar ar-4">
          <circle cx="50" cy="50" r="11" fill="none" stroke="#00ff9f" strokeWidth="1.3" opacity="0.6" strokeDasharray="40 30" />
        </g>
        <circle cx="50" cy="50" r="2.5" fill="#d4af37" />
      </svg>
      {label && <div className="astrolabe-label">{label}</div>}
    </div>
  );
}

function ProcessListItem({
  proc,
  active,
  onSelect,
  onKill,
  onDelete,
  onArchive,
  onRerun,
}: {
  proc: ProcessInfo;
  active: boolean;
  onSelect: () => void;
  onKill: () => void;
  onDelete: () => void;
  onArchive: () => void;
  onRerun: () => void;
}) {
  const [dragX, setDragX] = useState(0);
  const [revealed, setRevealed] = useState(false);
  const [archiving, setArchiving] = useState(false);
  const startRef = useRef<{ x: number; y: number; pointerId: number } | null>(null);
  const movedRef = useRef(false);

  const reset = () => {
    setDragX(0);
    setRevealed(false);
    setArchiving(false);
    startRef.current = null;
    movedRef.current = false;
  };

  const finishSwipe = (x: number) => {
    if (x <= -76) {
      setArchiving(true);
      window.setTimeout(onArchive, 90);
      return;
    }
    if (x >= 68) {
      setDragX(88);
      setRevealed(true);
      return;
    }
    reset();
  };

  return (
    <div
      className={`process-swipe ${revealed ? 'reveal-delete' : ''} ${archiving ? 'archiving' : ''}`}
    >
      <div className="swipe-action swipe-delete">
        <button
          className="swipe-action-button delete"
          aria-label="Delete command"
          title="Delete"
          onClick={(e) => {
            e.stopPropagation();
            onDelete();
          }}
        >
          <Trash2 size={17} aria-hidden="true" />
        </button>
      </div>
      <div className="swipe-action swipe-archive" aria-hidden="true" title="Archive">
        <Archive className="swipe-action-icon" size={18} />
      </div>
      <div
        className={`process-item ${active ? 'active' : ''}`}
        style={{ transform: `translateX(${dragX}px)` }}
        onPointerDown={(e) => {
          startRef.current = { x: e.clientX, y: e.clientY, pointerId: e.pointerId };
          movedRef.current = false;
          (e.currentTarget as HTMLElement).setPointerCapture(e.pointerId);
        }}
        onPointerMove={(e) => {
          const start = startRef.current;
          if (!start) return;
          const dx = e.clientX - start.x;
          const dy = e.clientY - start.y;
          if (Math.abs(dy) > Math.abs(dx) && Math.abs(dy) > 8) return;
          if (Math.abs(dx) > 8) movedRef.current = true;
          const clamped = Math.max(-112, Math.min(104, dx));
          setDragX(clamped);
          setRevealed(false);
        }}
        onPointerUp={(e) => {
          const start = startRef.current;
          if (!start) return;
          const dx = e.clientX - start.x;
          startRef.current = null;
          finishSwipe(dx);
        }}
        onPointerCancel={reset}
        onClick={() => {
          if (movedRef.current) {
            movedRef.current = false;
            return;
          }
          onSelect();
        }}
      >
        <div className="proc-main">
          <code>{proc.command.substring(0, 48)}</code>
          {proc.status === 'running' ? (
            <span className="run-dot">● running</span>
          ) : proc.exit_code !== null ? (
            <span className={`exit-pill ${proc.exit_code === 0 ? 'ok' : 'bad'}`}>
              {proc.exit_code === 0 ? '✓' : '✗'} {proc.exit_code}
            </span>
          ) : (
            <span className={`status ${proc.status}`}>{proc.status}</span>
          )}
        </div>
        <div className="proc-footer">
          <small>
            {new Date(proc.created_at).toLocaleTimeString()}
            {proc.duration_ms != null && ` · ${proc.duration_ms}ms`}
          </small>
          {proc.status === 'running' ? (
            <button
              className="btn btn-kill"
              onClick={(e) => {
                e.stopPropagation();
                onKill();
              }}
            >
              kill
            </button>
          ) : (
            <button
              className="btn btn-mini"
              onClick={(e) => {
                e.stopPropagation();
                onRerun();
              }}
              title="run this command again"
            >
              ↻ run
            </button>
          )}
        </div>
      </div>
    </div>
  );
}

function PtySessionListItem({
  session,
  active,
  onSelect,
  onClose,
}: {
  session: PtySessionInfo;
  active: boolean;
  onSelect: () => void;
  onClose: () => void;
}) {
  return (
    <div
      className={`process-item pty-process-item ${active ? 'active' : ''}`}
      onClick={onSelect}
    >
      <div className="proc-main">
        <code>{session.command.substring(0, 48)}</code>
        {session.running ? (
          <span className="run-dot">● live</span>
        ) : (
          <span className="status completed">exited</span>
        )}
      </div>
      <div className="proc-footer">
        <small>{new Date(session.created_at).toLocaleTimeString()} · pty</small>
        <button
          className={`btn ${session.running ? 'btn-kill' : 'btn-mini'}`}
          onClick={(e) => {
            e.stopPropagation();
            onClose();
          }}
          title={session.running ? 'kill session' : 'close session'}
        >
          {session.running ? 'kill' : 'close'}
        </button>
      </div>
    </div>
  );
}

function App() {
  const [processes, setProcesses] = useState<ProcessInfo[]>([]);
  const [command, setCommand] = useState('');
  const [loading, setLoading] = useState(false);
  const [selectedId, setSelectedId] = useState<string | null>(null);
  const [archivedIds, setArchivedIds] = useState<Set<string>>(() => new Set());
  const [deletedIds, setDeletedIds] = useState<Set<string>>(() => new Set());
  const [error, setError] = useState('');
  // Sidebar has three modes, cycled with the toggle or Cmd/Ctrl-B:
  //   'rail'     — narrow icon rail (default), CLI prompt has the rest
  //   'expanded' — full process list visible
  //   'hidden'   — fully collapsed for a distraction-free fullscreen terminal
  const [sidebarMode, setSidebarMode] = useState<'rail' | 'expanded' | 'hidden'>('rail');
  // Command history for Up/Down recall (most-recent last). historyIdx === null
  // means "editing a fresh line"; otherwise it indexes into history. Persisted
  // to localStorage so it survives app restarts (capped to the last 500).
  const [history, setHistory] = useState<string[]>(loadHistory);
  const [historyIdx, setHistoryIdx] = useState<number | null>(null);
  // Ctrl-R reverse-incremental search over history. searchMode toggles the
  // (reverse-i-search) prompt; searchQuery is the incremental needle;
  // searchNoMatch flags a query with no hit. searchPrevRef remembers the line
  // to restore if the search is cancelled.
  const [searchMode, setSearchMode] = useState(false);
  const [searchQuery, setSearchQuery] = useState('');
  const [searchNoMatch, setSearchNoMatch] = useState(false);
  const searchPrevRef = useRef('');
  const searchIdxRef = useRef<number | null>(null);
  const [showSettings, setShowSettings] = useState(false);
  const [showHelp, setShowHelp] = useState(false);
  const [session, setSession] = useState<AuthSession | null>(loadStoredSession);
  const sessionRef = useRef<AuthSession | null>(session);
  sessionRef.current = session;
  const initialMoonshineSettings = loadMoonshineSettings();
  const moonshineSettingsRef = useRef<MoonshineSettings>(initialMoonshineSettings);
  // Interactive PTY sessions (claude, codex, vim, ...). They are first-class
  // sidebar slices, so switching away does not discard the live session.
  const [ptySessions, setPtySessions] = useState<PtySessionInfo[]>([]);
  // Active PTY session. When set, the right pane remounts a terminal against
  // the backend PTY buffer for that slice.
  const [ptyId, setPtyId] = useState<string | null>(null);
  // VM tabs: each booted .iso/.img runs as its own PTY session (boot-iso →
  // Virtualization.framework) and KEEPS RUNNING in the background even when its
  // tab isn't on screen. `activeVmId === null` shows the normal shell/output.
  const [vmSessions, setVmSessions] = useState<{ id: string; label: string }[]>([]);
  const [activeVmId, setActiveVmId] = useState<string | null>(null);
  // Optional vim-style editing on the command line (off by default). When on,
  // the prompt starts in normal mode; i/a/etc enter insert, Esc returns.
  const [vimEnabled, setVimEnabled] = useState(
    () => initialMoonshineSettings.vimEnabled
  );
  const [vimMode, setVimMode] = useState<'normal' | 'insert'>('insert');
  const [agentGenTokens, setAgentGenTokensState] = useState(
    () => initialMoonshineSettings.agentGenTokens
  );
  const [accentColor, setAccentColorState] = useState(
    () => initialMoonshineSettings.accentColor
  );
  // Tab-completion candidates for the current token (shown below the prompt).
  const [completions, setCompletions] = useState<string[]>([]);
  const [completeIdx, setCompleteIdx] = useState(0);
  // URL-bar browser: when the prompt is a URL, the right pane shows the fetched
  // page text. history/historyIdx give back/forward navigation.
  const [browseUrl, setBrowseUrl] = useState<string | null>(null);
  const [browseText, setBrowseText] = useState('');
  const [browseLinks, setBrowseLinks] = useState<{ text: string; href: string }[]>([]);
  const [browseLoading, setBrowseLoading] = useState(false);
  const [browseCached, setBrowseCached] = useState(false);
  const [browseMs, setBrowseMs] = useState(0);
  const [cwd, setCwd] = useState('');
  const [branch, setBranch] = useState('');
  const [browseHist, setBrowseHist] = useState<string[]>([]);
  const [browseHistIdx, setBrowseHistIdx] = useState(-1);
  // Easter egg: double-tap Shift on the empty pane reveals the cummings poem.
  const [easterEgg, setEasterEgg] = useState(false);
  // Zoom level (Cmd-+/-/0), persisted. 1.0 = default; clamped 0.6–2.0.
  const [zoom, setZoom] = useState<number>(() => initialMoonshineSettings.zoom);
  const inputRef = useRef<HTMLTextAreaElement>(null);
  // Mirror of `processes` for event handlers that close over stale state (the
  // global key handler only re-binds on a couple of deps).
  const processesRef = useRef<ProcessInfo[]>([]);
  processesRef.current = processes;
  const ptySessionsRef = useRef<PtySessionInfo[]>([]);
  ptySessionsRef.current = ptySessions;
  const selectedIdRef = useRef<string | null>(null);
  selectedIdRef.current = selectedId;
  const ptyIdRef = useRef<string | null>(null);
  ptyIdRef.current = ptyId;

  // Apply zoom to the whole UI via the CSS `zoom` property (scales fixed-px
  // layout uniformly, unlike root font-size) and persist it.
  useEffect(() => {
    const applyZoom = () => {
      const mobileViewport =
        window.matchMedia('(max-width: 720px), (pointer: coarse)').matches;
      (document.body.style as unknown as { zoom: string }).zoom = mobileViewport
        ? '1'
        : String(zoom);
    };
    applyZoom();
    window.addEventListener('resize', applyZoom);
    localStorage.setItem('moonshine.zoom', String(zoom));
    return () => window.removeEventListener('resize', applyZoom);
  }, [zoom]);

  useEffect(() => {
    const setViewportHeight = () => {
      const height = window.visualViewport?.height ?? window.innerHeight;
      document.documentElement.style.setProperty('--viewport-height', `${height}px`);
    };
    setViewportHeight();
    window.visualViewport?.addEventListener('resize', setViewportHeight);
    window.visualViewport?.addEventListener('scroll', setViewportHeight);
    window.addEventListener('resize', setViewportHeight);
    window.addEventListener('orientationchange', setViewportHeight);
    return () => {
      window.visualViewport?.removeEventListener('resize', setViewportHeight);
      window.visualViewport?.removeEventListener('scroll', setViewportHeight);
      window.removeEventListener('resize', setViewportHeight);
      window.removeEventListener('orientationchange', setViewportHeight);
    };
  }, []);
  const applyMoonshineSettings = (settings: MoonshineSettings) => {
    const normalized = saveMoonshineSettings(settings);
    moonshineSettingsRef.current = normalized;
    setVimEnabled(normalized.vimEnabled);
    setVimMode(normalized.vimEnabled ? 'normal' : 'insert');
    setAgentGenTokensState(normalized.agentGenTokens);
    setAccentColorState(normalized.accentColor);
    setZoom(normalized.zoom);
    void checkpointMoonshineSettings(sessionRef.current, normalized);
  };

  const updateMoonshineSettings = (partial: Partial<MoonshineSettings>) => {
    applyMoonshineSettings(
      normalizeMoonshineSettings({
        ...moonshineSettingsRef.current,
        ...partial,
      })
    );
  };

  const bumpZoom = (delta: number) => {
    updateMoonshineSettings({
      zoom: normalizeZoom(moonshineSettingsRef.current.zoom + delta),
    });
  };

  const setVim = (on: boolean) => {
    updateMoonshineSettings({ vimEnabled: on });
  };

  const setAgentGenTokens = (value: number) => {
    updateMoonshineSettings({ agentGenTokens: normalizeAgentGenTokens(value) });
  };

  const setAccentColor = (value: string) => {
    updateMoonshineSettings({ accentColor: normalizeAccentColor(value) });
  };

  useEffect(() => {
    if (session) {
      void checkpointMoonshineSettings(session, moonshineSettingsRef.current);
    }
  }, [session]);

  useEffect(() => {
    document.documentElement.style.setProperty('--accent-green', accentColor);
  }, [accentColor]);

  const cycleSidebar = () =>
    setSidebarMode((m) => (m === 'rail' ? 'expanded' : m === 'expanded' ? 'hidden' : 'rail'));

  useEffect(() => {
    loadProcesses();
    const interval = setInterval(loadProcesses, 1000);
    return () => clearInterval(interval);
  }, []);

  // App-global hotkeys (the things that make a terminal fast to drive):
  //   Cmd/Ctrl-B        cycle sidebar (rail → expanded → hidden)
  //   Cmd/Ctrl-,        toggle settings
  //   Cmd/Ctrl-Enter    focus the prompt
  //   Cmd-Ctrl-F        native fullscreen (also F11)
  //   Esc               leave settings / exit fullscreen
  useEffect(() => {
    const toggleFullscreen = async () => {
      try {
        const { getCurrentWindow } = await import('@tauri-apps/api/window');
        const w = getCurrentWindow();
        const isFs = await w.isFullscreen();
        await w.setFullscreen(!isFs);
      } catch (e) {
        console.error('fullscreen toggle failed:', e);
      }
    };

    let lastShift = 0;
    const onGlobalKey = (e: KeyboardEvent) => {
      const mod = e.metaKey || e.ctrlKey;
      // Ctrl-C interrupts a running command (restarts the wedged REPL). Only
      // when something is actually running — otherwise fall through so the
      // input's readline Ctrl-C (clear line) and text-copy keep working.
      if (e.ctrlKey && !e.metaKey && (e.key === 'c' || e.key === 'C')) {
        if (processesRef.current.some((p) => p.status === 'running')) {
          e.preventDefault();
          invoke('interrupt_command')
            .then(() => loadProcesses())
            .catch(() => {});
          return;
        }
      }
      // Double-tap Shift (within 400ms, no other modifier) → cummings easter egg.
      if (e.key === 'Shift' && !e.metaKey && !e.ctrlKey && !e.altKey) {
        const now = e.timeStamp;
        if (now - lastShift < 400) {
          setEasterEgg((v) => !v);
          lastShift = 0;
        } else {
          lastShift = now;
        }
        return;
      }
      // Zoom: Cmd/Ctrl +/=/- and 0 to reset.
      if (mod && (e.key === '=' || e.key === '+')) {
        e.preventDefault();
        bumpZoom(0.1);
        return;
      } else if (mod && e.key === '-') {
        e.preventDefault();
        bumpZoom(-0.1);
        return;
      } else if (mod && e.key === '0') {
        e.preventDefault();
        updateMoonshineSettings({ zoom: 1 });
        return;
      }
      if (mod && e.key.toLowerCase() === 'b') {
        e.preventDefault();
        cycleSidebar();
      } else if (mod && e.key === ',') {
        e.preventDefault();
        setShowSettings((s) => !s);
      } else if (mod && e.key === '/') {
        // Cmd/Ctrl-/ — toggle the keyboard-shortcuts cheat sheet.
        e.preventDefault();
        setShowHelp((h) => !h);
      } else if (mod && e.key.toLowerCase() === 'k') {
        // Cmd/Ctrl-K — clear the finished command history (Terminal/iTerm idiom).
        e.preventDefault();
        clearHistory();
      } else if (mod && (e.key === '[' || e.key === ']')) {
        // Cmd/Ctrl-[ / ] — step the selected command output older / newer
        // (back/forward), so you can review past outputs and PTY slices without
        // the mouse.
        e.preventDefault();
        const slices = [
          ...ptySessionsRef.current.map((session) => ({ kind: 'pty' as const, id: session.id })),
          ...processesRef.current.map((proc) => ({ kind: 'repl' as const, id: proc.id })),
        ];
        if (slices.length === 0) return;
        const activeId = ptyIdRef.current ?? selectedIdRef.current ?? '';
        const cur = slices.findIndex((slice) => slice.id === activeId);
        // The list is newest-first: '[' (older) moves down it, ']' (newer) up.
        const next =
          cur < 0
            ? 0
            : e.key === '['
            ? Math.min(slices.length - 1, cur + 1)
            : Math.max(0, cur - 1);
        setBrowseUrl(null);
        setShowSettings(false);
        setShowHelp(false);
        const slice = slices[next];
        if (!slice) return;
        if (slice.kind === 'pty') {
          setPtyId(slice.id);
          setSelectedId(null);
        } else {
          setPtyId(null);
          setSelectedId(slice.id);
        }
      } else if (mod && e.key === 'Enter') {
        e.preventDefault();
        inputRef.current?.focus();
      } else if ((e.metaKey && e.ctrlKey && e.key.toLowerCase() === 'f') || e.key === 'F11') {
        e.preventDefault();
        toggleFullscreen();
      } else if (e.key === 'Escape') {
        if (showHelp) setShowHelp(false);
        else if (easterEgg) setEasterEgg(false);
        else if (showSettings) setShowSettings(false);
      }
    };

    window.addEventListener('keydown', onGlobalKey);
    return () => window.removeEventListener('keydown', onGlobalKey);
  }, [showSettings, easterEgg, showHelp]);

  const loadProcesses = async () => {
    try {
      const procs: ProcessInfo[] = await invoke('list_processes');
      setProcesses(procs);
    } catch (e) {
      console.error('Failed to list processes:', e);
    }
  };

  // Clear the finished command history (declutter + reclaim the stored output
  // memory). A running command is kept by the backend.
  const clearHistory = () => {
    invoke('clear_history')
      .then(() => {
        setPtySessions((sessions) => sessions.filter((session) => session.running));
        setPtyId((cur) =>
          cur && ptySessionsRef.current.some((session) => session.id === cur && session.running)
            ? cur
            : null
        );
        setSelectedId(null);
        loadProcesses();
      })
      .catch(() => {});
  };

  const spawn = async (e: React.FormEvent, override?: string) => {
    e.preventDefault();
    const raw = override !== undefined ? override : command;
    const submitted = (agenticModalityCommand(raw) ?? raw.trim()) || STARTER_COMMAND;
    // Record in history (skip consecutive duplicates), reset recall cursor.
    const nextHistory =
      history[history.length - 1] === submitted ? history : saveHistory([...history, submitted]);
    setHistory(nextHistory);
    setHistoryIdx(null);
    // Vim convention: a fresh prompt returns to normal mode after submit.
    if (vimEnabled) setVimMode('normal');

    // URL-bar route: a lone URL / aeon:// / bare domain is browsed via the
    // backend `browse` command (crisp `wall --text` DOM text), so the prompt
    // doubles as a URL bar.
    const url = asBrowsableUrl(submitted);
    if (url) {
      setCommand('');
      // Push onto browser history (truncate any forward entries).
      setBrowseHist((h) => {
        const base = h.slice(0, browseHistIdx + 1);
        return [...base, url];
      });
      setBrowseHistIdx((i) => i + 1);
      await navigate(url);
      return;
    }

    // A VM target (.iso/.img, `boot-iso …`, or `vm …`) boots as its own tab.
    const vt = vmTarget(submitted);
    if (vt) {
      await bootVm(vt);
      return;
    }

    // Interactive programs (claude, vim, python, ssh, ...) need a real TTY, so
    // they run in a PTY pane instead of the fast non-interactive REPL.
    if (isInteractiveCommand(submitted)) {
      setLoading(true);
      setError('');
      try {
        const ptyCommand = ptyExecutionCommand(submitted);
        // 80x24 is a safe default; the terminal reports its real size on mount.
        const id: string = await invoke('spawn_pty', {
          command: ptyCommand,
          cols: 80,
          rows: 24,
          cwd,
          agentGenTokens,
          history: nextHistory,
        });
        setPtySessions((sessions) => [
          {
            id,
            command: submitted,
            executedCommand: ptyCommand,
            created_at: new Date().toISOString(),
            running: true,
          },
          ...sessions,
        ]);
        setPtyId(id);
        setSelectedId(null);
        setActiveVmId(null);
        setBrowseUrl(null);
        setShowSettings(false);
        setCommand('');
      } catch (err) {
        setError(String(err));
      } finally {
        setLoading(false);
      }
      return;
    }

    setLoading(true);
    setError('');
    try {
      const proc: ProcessInfo = await invoke('spawn_shell', {
        command: submitted,
        agentGenTokens,
      });
      setCommand('');
      setSelectedId(proc.id);
      // A new command yields the right pane: drop any (possibly exited) PTY
      // pane so this command's output is shown instead of staying hidden
      // behind it (ptyId wins the render otherwise), and leave the VM tabs (they
      // keep running) so the shell output is visible.
      setPtyId(null);
      setActiveVmId(null);
      await loadProcesses();
      inputRef.current?.focus();
      // Refresh the cwd indicator — the REPL serializes, so this pwd runs
      // after the just-submitted command and reflects any `cd`.
      refreshCwd();
    } catch (err) {
      setError(String(err));
    } finally {
      setLoading(false);
    }
  };

  const runAgenticModality = (cmd: string) => {
    setCommand(cmd);
    void spawn({ preventDefault() {} } as React.FormEvent, cmd);
  };

  // Query the REPL's working directory + git branch for the prompt indicator.
  const refreshCwd = () => {
    invoke<{ cwd: string; branch: string }>('repl_cwd', { agentGenTokens })
      .then((c) => {
        setCwd(c?.cwd || '');
        setBranch(c?.branch || '');
      })
      .catch(() => {});
  };

  // Warm the guarded REPL at startup. It spawns lazily otherwise, so the very
  // first command would pay the full ~700ms guard cost (delegate + run + UCAN
  // crypto) while every later command is ~5ms. Kicking the spawn off on mount —
  // in the background while the user reads the UI — makes the first command
  // warm too. (repl_cwd inits the REPL and also fills the cwd bar immediately.)
  useEffect(() => {
    refreshCwd();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  // Reflect context in the window title (like a real terminal's OSC title): the
  // running PTY program, else the browsed host, else the working directory.
  // Lets you pick the right window when several are open.
  const activePty = ptyId ? ptySessions.find((session) => session.id === ptyId) ?? null : null;
  const activePtyCommand = activePty?.command ?? '';
  const activePtyRunning = activePty?.running ?? false;

  useEffect(() => {
    let title = 'Moonshine';
    if (activePtyRunning) {
      title = `moonshine — ${activePtyCommand.split(/\s+/)[0]}`;
    } else if (browseUrl) {
      let host = browseUrl;
      try {
        host = new URL(browseUrl).host || browseUrl;
      } catch {
        /* keep raw */
      }
      title = `moonshine — ${host}`;
    } else if (formatCwdLabel(cwd)) {
      title = `moonshine — ${formatCwdLabel(cwd)}`;
      if (branch) title += ` (${branch})`;
    }
    import('@tauri-apps/api/window')
      .then(({ getCurrentWindow }) => getCurrentWindow().setTitle(title))
      .catch(() => {});
  }, [cwd, branch, activePtyCommand, activePtyRunning, browseUrl]);

  // Fetch a URL's crisp text and show it in the browser pane. The prompt line
  // and the URL bar are ONE thing: while browsing, the prompt shows the current
  // URL (editable to navigate elsewhere), so there's no separate address bar.
  const navigate = async (url: string, force = false) => {
    setBrowseUrl(url);
    setBrowseLoading(true);
    setBrowseText('');
    setBrowseLinks([]);
    setSelectedId(null);
    setPtyId(null);
    setShowSettings(false);
    setError('');
    setCommand(url); // collapse: the prompt becomes the address bar
    const t0 = performance.now();
    try {
      const res: { text: string; links: { text: string; href: string }[]; from_cache: boolean } =
        await invoke('browse', { url, force });
      setBrowseText(res.text);
      setBrowseLinks(res.links || []);
      setBrowseCached(!!res.from_cache);
      setBrowseMs(Math.round(performance.now() - t0));
    } catch (err) {
      setBrowseText('');
      setBrowseLinks([]);
      setError(String(err));
    } finally {
      setBrowseLoading(false);
      inputRef.current?.focus();
    }
  };

  const browseBack = () => {
    if (browseHistIdx <= 0) return;
    const i = browseHistIdx - 1;
    setBrowseHistIdx(i);
    navigate(browseHist[i]);
  };
  const browseForward = () => {
    if (browseHistIdx >= browseHist.length - 1) return;
    const i = browseHistIdx + 1;
    setBrowseHistIdx(i);
    navigate(browseHist[i]);
  };

  // Open a URL found in command output in the browser pane (same path as a
  // clicked page link: truncate any forward history, then navigate).
  const openFromOutput = (url: string) => {
    setBrowseHist((h) => [...h.slice(0, browseHistIdx + 1), url]);
    setBrowseHistIdx((i) => i + 1);
    navigate(url);
  };

  // Set value and move the caret in one shot (state + DOM, since React can't
  // place the caret declaratively).
  const setLine = (value: string, caret?: number) => {
    setCommand(value);
    const pos = caret === undefined ? value.length : caret;
    requestAnimationFrame(() => {
      const el = inputRef.current;
      if (el) el.setSelectionRange(pos, pos);
    });
  };

  const applyMoonshineLink = (uri: string) => {
    try {
      const url = new URL(uri);
      if (url.protocol !== 'moonshine:') {
        setLine(uri);
        return;
      }
      const commandParam = url.searchParams.get('command') || url.searchParams.get('cmd');
      const browseParam = url.searchParams.get('url') || url.searchParams.get('href');
      if ((url.hostname === 'browse' || url.hostname === 'open') && browseParam) {
        const nextUrl = asBrowsableUrl(browseParam) || browseParam;
        setBrowseHist((h) => [...h.slice(0, browseHistIdx + 1), nextUrl]);
        setBrowseHistIdx((i) => i + 1);
        navigate(nextUrl);
        return;
      }
      if (commandParam) {
        setBrowseUrl(null);
        setLine(commandParam);
        return;
      }
      if (url.hostname === 'run' && url.pathname.length > 1) {
        setBrowseUrl(null);
        setLine(decodeURIComponent(url.pathname.slice(1)));
        return;
      }
      setBrowseUrl(null);
      setLine(uri);
    } catch {
      setLine(uri);
    }
  };

  useEffect(() => {
    const unlisten = listen<{ uri?: string | string[] }>('moonshine://deep-link', (event) => {
      const uri = event.payload?.uri;
      const first = Array.isArray(uri) ? uri[0] : uri;
      if (typeof first === 'string' && first.trim()) {
        applyMoonshineLink(first);
      }
    });
    return () => {
      unlisten.then((off) => off()).catch(() => {});
    };
  }, [browseHistIdx]);

  // Readline-style line editing + history, the shortcuts that make a terminal
  // fast: Up/Down history, Ctrl-A/E (home/end), Ctrl-K (kill to end),
  // Ctrl-U (kill to start), Ctrl-W (delete word back), Ctrl-L (clear screen).
  // Vim normal-mode motions/edits on the command line. Returns true if the key
  // was handled (so the default + readline paths are skipped).
  const handleVimNormal = (e: React.KeyboardEvent<HTMLTextAreaElement>): boolean => {
    const el = e.currentTarget;
    const caret = el.selectionStart ?? 0;
    const k = e.key;
    const move = (pos: number) => el.setSelectionRange(pos, pos);
    // Find word boundaries from the caret.
    const nextWord = () => {
      const m = command.slice(caret).match(/^\s*\S+\s*/);
      return caret + (m ? m[0].length : 0);
    };
    const prevWord = () => {
      const left = command.slice(0, caret).replace(/\s*\S+\s*$/, '');
      return left.length;
    };
    switch (k) {
      case 'i': e.preventDefault(); setVimMode('insert'); return true;
      case 'a': e.preventDefault(); move(Math.min(caret + 1, command.length)); setVimMode('insert'); return true;
      case 'I': e.preventDefault(); move(0); setVimMode('insert'); return true;
      case 'A': e.preventDefault(); move(command.length); setVimMode('insert'); return true;
      case 'h': case 'ArrowLeft': e.preventDefault(); move(Math.max(0, caret - 1)); return true;
      case 'l': case 'ArrowRight': e.preventDefault(); move(Math.min(command.length, caret + 1)); return true;
      case '0': e.preventDefault(); move(0); return true;
      case '$': e.preventDefault(); move(command.length); return true;
      case 'w': e.preventDefault(); move(nextWord()); return true;
      case 'b': e.preventDefault(); move(prevWord()); return true;
      case 'x': e.preventDefault(); setLine(command.slice(0, caret) + command.slice(caret + 1), caret); return true;
      case 'D': e.preventDefault(); setLine(command.slice(0, caret), caret); return true;
      case 'C': e.preventDefault(); setLine(command.slice(0, caret), caret); setVimMode('insert'); return true;
      case 'd': // crude dd: clear the whole line
        e.preventDefault(); setLine('', 0); return true;
      case 'k': // history prev (vim up)
        e.preventDefault();
        if (history.length) {
          const idx = historyIdx === null ? history.length - 1 : Math.max(0, historyIdx - 1);
          setHistoryIdx(idx); setLine(history[idx]);
        }
        return true;
      case 'j': // history next (vim down)
        e.preventDefault();
        if (historyIdx !== null) {
          const n = historyIdx + 1;
          if (n >= history.length) { setHistoryIdx(null); setLine(''); }
          else { setHistoryIdx(n); setLine(history[n]); }
        }
        return true;
      case 'Enter': return false; // let submit happen from normal mode too
      default:
        // Swallow other printable keys in normal mode (don't insert text).
        if (k.length === 1 && !e.ctrlKey && !e.metaKey) { e.preventDefault(); return true; }
        return false;
    }
  };

  // Replace the token under the caret with `value`, keeping the rest of the
  // line, and place the caret at the end of the inserted token.
  const replaceToken = (value: string) => {
    const el = inputRef.current;
    const caret = el?.selectionStart ?? command.length;
    const left = command.slice(0, caret);
    const right = command.slice(caret);
    const tokenStart = Math.max(left.lastIndexOf(' '), left.lastIndexOf('\t')) + 1;
    const newLeft = command.slice(0, tokenStart) + value;
    setLine(newLeft + right, newLeft.length);
  };

  // Tab-complete the current token: first Tab fetches candidates; if exactly
  // one, apply it; if many, apply the first and let repeated Tab cycle.
  const doComplete = async (e: React.KeyboardEvent<HTMLTextAreaElement>) => {
    e.preventDefault();
    const caret = e.currentTarget.selectionStart ?? command.length;
    // If we already have a candidate list, cycle to the next one.
    if (completions.length > 1) {
      const next = (completeIdx + 1) % completions.length;
      setCompleteIdx(next);
      replaceToken(completions[next]);
      return;
    }
    try {
      const cands: string[] = await invoke('complete', { input: command, caret, cwd });
      if (cands.length === 1) {
        replaceToken(cands[0]);
        setCompletions([]);
      } else if (cands.length > 1) {
        // Extend the token to the longest common prefix (so the caret stays put
        // and nothing is guessed for you), then show the list. Repeated Tab then
        // cycles through the full candidates (menu-complete).
        const left = command.slice(0, caret);
        const tokenStart = Math.max(left.lastIndexOf(' '), left.lastIndexOf('\t')) + 1;
        const token = left.slice(tokenStart);
        const lcp = longestCommonPrefix(cands);
        if (lcp.length > token.length) replaceToken(lcp);
        setCompletions(cands);
        setCompleteIdx(-1);
      }
    } catch (err) {
      console.error('completion failed:', err);
    }
  };

  // Most-recent-first scan of history for the newest entry containing `query`,
  // starting at index `from` and going older. Returns -1 if none.
  const findMatch = (query: string, from: number): number => {
    if (!query) return -1;
    const q = query.toLowerCase();
    for (let i = Math.min(from, history.length - 1); i >= 0; i--) {
      if (history[i].toLowerCase().includes(q)) return i;
    }
    return -1;
  };

  // Apply a reverse-search query: show the newest match (or older than `before`
  // when cycling), update the prompt line + no-match flag.
  const applySearch = (query: string, before: number) => {
    setSearchQuery(query);
    const m = findMatch(query, before);
    searchIdxRef.current = m === -1 ? null : m;
    if (m !== -1) {
      setSearchNoMatch(false);
      setLine(history[m]);
    } else {
      setSearchNoMatch(!!query);
    }
  };

  const exitSearch = (accept: boolean) => {
    setSearchMode(false);
    setSearchQuery('');
    setSearchNoMatch(false);
    searchIdxRef.current = null;
    if (!accept) setLine(searchPrevRef.current);
  };

  const onKeyDown = (e: React.KeyboardEvent<HTMLTextAreaElement>) => {
    const el = e.currentTarget;
    const caret = el.selectionStart ?? command.length;
    const ctrl = e.ctrlKey && !e.metaKey;

    // Ctrl-R: enter reverse-i-search, or cycle to the next older match.
    if (ctrl && (e.key === 'r' || e.key === 'R')) {
      e.preventDefault();
      if (!searchMode) {
        searchPrevRef.current = command;
        setSearchMode(true);
        setSearchQuery('');
        setSearchNoMatch(false);
        searchIdxRef.current = null;
        return;
      }
      // Already searching: find an older match than the current index.
      const start = searchIdxRef.current === null ? history.length - 1 : searchIdxRef.current - 1;
      const m = findMatch(searchQuery, start);
      if (m !== -1) {
        searchIdxRef.current = m;
        setSearchNoMatch(false);
        setLine(history[m]);
      } else {
        setSearchNoMatch(!!searchQuery);
      }
      return;
    }

    // While searching, keys edit the query / accept / cancel.
    if (searchMode) {
      if (e.key === 'Escape' || (ctrl && (e.key === 'g' || e.key === 'c'))) {
        e.preventDefault();
        exitSearch(false);
        return;
      }
      if (e.key === 'Enter') {
        // Accept the match (leave it in the input) and fall through so the
        // normal Enter handler below runs it.
        exitSearch(true);
      } else if (e.key === 'Backspace') {
        e.preventDefault();
        applySearch(searchQuery.slice(0, -1), history.length - 1);
        return;
      } else if (e.key.length === 1 && !ctrl && !e.metaKey && !e.altKey) {
        e.preventDefault();
        applySearch(searchQuery + e.key, history.length - 1);
        return;
      } else if (e.key === 'Shift' || e.key === 'Control' || e.key === 'Meta' || e.key === 'Alt') {
        // A bare modifier press shouldn't end the search.
        return;
      } else {
        // Any other key (arrows, Tab, ctrl-edits) accepts the match and falls
        // through to its normal handler below.
        exitSearch(true);
      }
    }

    // Multi-line: Enter submits, Shift-Enter inserts a newline (textarea
    // default is newline, so we invert it for the common single-line case).
    if (e.key === 'Enter' && !e.shiftKey) {
      e.preventDefault();
      // Submit via the form path (reuses URL/PTY/REPL routing + history).
      spawn(e as unknown as React.FormEvent);
      return;
    }

    if (e.key === 'Tab') {
      doComplete(e);
      return;
    }
    // Any non-Tab key invalidates the current completion cycle.
    if (completions.length) setCompletions([]);

    // Vim mode: Esc leaves insert -> normal; normal-mode keys are motions/edits.
    if (vimEnabled) {
      if (e.key === 'Escape') {
        e.preventDefault();
        setVimMode('normal');
        return;
      }
      if (vimMode === 'normal') {
        if (handleVimNormal(e)) return;
      }
    }

    if (e.key === 'ArrowUp' && !e.altKey) {
      // In a multi-line command, Up moves between lines; only recall older
      // history when the caret is already on the first line (zsh/fish behavior).
      if (command.slice(0, caret).includes('\n')) return;
      if (history.length === 0) return;
      e.preventDefault();
      const idx = historyIdx === null ? history.length - 1 : Math.max(0, historyIdx - 1);
      setHistoryIdx(idx);
      setLine(history[idx]);
    } else if (e.key === 'ArrowDown' && !e.altKey) {
      // Symmetric: Down moves between lines unless the caret is on the last line.
      if (command.slice(caret).includes('\n')) return;
      if (historyIdx === null) return;
      e.preventDefault();
      const next = historyIdx + 1;
      if (next >= history.length) {
        setHistoryIdx(null);
        setLine('');
      } else {
        setHistoryIdx(next);
        setLine(history[next]);
      }
    } else if (ctrl && e.key === 'a') {
      // Ctrl-A: move to start of line.
      e.preventDefault();
      el.setSelectionRange(0, 0);
    } else if (ctrl && e.key === 'e') {
      // Ctrl-E: move to end of line.
      e.preventDefault();
      el.setSelectionRange(command.length, command.length);
    } else if (ctrl && e.key === 'k') {
      // Ctrl-K: kill from caret to end.
      e.preventDefault();
      setLine(command.slice(0, caret), caret);
    } else if (ctrl && e.key === 'u') {
      // Ctrl-U: kill from start to caret.
      e.preventDefault();
      setLine(command.slice(caret), 0);
    } else if (ctrl && e.key === 'w') {
      // Ctrl-W: delete the word before the caret.
      e.preventDefault();
      const left = command.slice(0, caret);
      const trimmed = left.replace(/\s*\S+\s*$/, '');
      setLine(trimmed + command.slice(caret), trimmed.length);
    } else if (ctrl && e.key === 'l') {
      // Ctrl-L: clear the visible output (deselect current process).
      e.preventDefault();
      setSelectedId(null);
    } else if (ctrl && e.key === 'c') {
      // Ctrl-C: abandon the current line (don't kill the app).
      e.preventDefault();
      setHistoryIdx(null);
      setLine('');
    }
  };

  const kill = async (id: string) => {
    try {
      await invoke('kill_process', { processId: id });
      await loadProcesses();
      if (selectedId === id) setSelectedId(null);
    } catch (err) {
      setError(String(err));
    }
  };

  const deleteProcess = async (id: string) => {
    setDeletedIds((ids) => new Set(ids).add(id));
    if (selectedId === id) setSelectedId(null);
    try {
      await invoke('kill_process', { processId: id });
      await loadProcesses();
    } catch (err) {
      setDeletedIds((ids) => {
        const next = new Set(ids);
        next.delete(id);
        return next;
      });
      setError(String(err));
    }
  };

  const archiveProcess = (id: string) => {
    setArchivedIds((ids) => new Set(ids).add(id));
    if (selectedId === id) setSelectedId(null);
  };

  const selectPtySession = (id: string) => {
    setPtyId(id);
    setSelectedId(null);
    setActiveVmId(null);
    setBrowseUrl(null);
    setShowSettings(false);
    setShowHelp(false);
  };

  const closePtySession = (id: string) => {
    invoke('pty_kill', { sessionId: id }).catch(() => {});
    setPtySessions((sessions) => sessions.filter((session) => session.id !== id));
    setPtyId((cur) => (cur === id ? null : cur));
  };

  const markPtyExited = (id: string) => {
    setPtySessions((sessions) =>
      sessions.map((session) =>
        session.id === id ? { ...session, running: false } : session
      )
    );
  };

  const focusNewSlice = () => {
    if (sidebarMode === 'hidden') setSidebarMode('rail');
    setSelectedId(null);
    setActiveVmId(null);
    setPtyId(null);
    setBrowseUrl(null);
    setTimeout(() => {
      inputRef.current?.focus();
      inputRef.current?.select();
    }, 0);
  };

  // Re-run a past command (from the process list) through the full router.
  const rerun = (cmd: string) => {
    setCommand(cmd);
    spawn({ preventDefault() {} } as React.FormEvent, cmd);
  };

  // Boot a VM (an .iso/.img) as a new persistent terminal tab and focus it. The
  // backend PTY session keeps the VM running even when another tab is on screen.
  const bootVm = async (target: string) => {
    const resolved = resolveBootUrl(target);
    if (!resolved) return;
    try {
      const id: string = await invoke('spawn_vm', {
        command: `boot-iso ${shQuote(resolved)}`,
        cols: 80,
        rows: 24,
        cwd,
      });
      setVmSessions((s) => [...s, { id, label: baseName(target) }]);
      setActiveVmId(id);
      setCommand('');
      setSelectedId(null);
      setPtyId(null);
      setShowSettings(false);
    } catch (err) {
      setError(String(err));
    }
  };

  // Switch to a VM tab (re-fits xterm to the pane after it mounts).
  const switchVm = (id: string) => {
    setActiveVmId(id);
    setTimeout(() => window.dispatchEvent(new Event('resize')), 0);
  };

  // Close a VM tab: shut the VM down (kill its PTY) and drop the tab.
  const closeVm = (id: string) => {
    invoke('pty_kill', { sessionId: id }).catch(() => {});
    setVmSessions((s) => s.filter((v) => v.id !== id));
    setActiveVmId((cur) => (cur === id ? null : cur));
  };

  // OS "Open With → Moonshine" on a .iso (Finder/launch/second-instance) boots
  // it as a VM tab. The Rust side emits `open-iso` with the resolved paths.
  useEffect(() => {
    let un: (() => void) | null = null;
    (async () => {
      un = await listen<{ paths: string[] }>('open-iso', (e) => {
        for (const p of e.payload.paths || []) void bootVm(p);
      });
    })();
    return () => {
      if (un) un();
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const expanded = sidebarMode === 'expanded';
  const visibleProcesses = processes.filter(
    (proc) => !archivedIds.has(proc.id) && !deletedIds.has(proc.id)
  );
  const visibleProcessCount = visibleProcesses.length + ptySessions.length;
  const cwdLabel = formatCwdLabel(cwd);
  return (
    <div className={`app ${sidebarMode === 'hidden' ? 'sidebar-hidden' : ''}`}>
      {sidebarMode !== 'hidden' && (
      <div className={`pane left ${expanded ? '' : 'collapsed'}`}>
        <div className="sidebar-header">
          <button
            className="btn btn-toggle"
            onClick={cycleSidebar}
            title={expanded ? 'collapse to rail' : 'expand processes'}
          >
            {expanded ? '‹' : '☰'}
          </button>
          {expanded && (
            <span
              className="sidebar-title"
              title="processes"
              style={{ display: 'inline-flex', alignItems: 'center', gap: 4 }}
            >
              <Activity size={14} aria-label="processes" /> ({visibleProcessCount})
            </span>
          )}
          {expanded && visibleProcessCount > 0 && (
            <button
              className="btn btn-toggle btn-clear"
              onClick={clearHistory}
              title="clear finished history — reclaims output memory (Cmd/Ctrl-K)"
            >
              clear
            </button>
          )}
          <button
            className="btn btn-toggle btn-new-slice"
            onClick={focusNewSlice}
            title="new slice: focus command palette"
          >
            +
          </button>
          {expanded && (
            <button
              className={`btn btn-toggle btn-help ${showHelp ? 'active' : ''}`}
              onClick={() => setShowHelp((h) => !h)}
              title="keyboard shortcuts (Cmd/Ctrl-/)"
            >
              ?
            </button>
          )}
          {expanded && (
            <button
              className={`btn btn-toggle btn-settings ${showSettings ? 'active' : ''} ${session ? 'has-avatar' : ''}`}
              onClick={() => setShowSettings((s) => !s)}
              title={session ? `${session.identity.name || session.identity.email || 'signed in'} — settings (Cmd/Ctrl-,)` : 'settings (Cmd/Ctrl-,)'}
            >
              {session?.identity.picture ? (
                <img className="avatar" src={session.identity.picture} alt="" />
              ) : session ? (
                <span className="avatar avatar-initial">
                  {(session.identity.name || session.identity.email || '?').charAt(0).toUpperCase()}
                </span>
              ) : (
                '⚙'
              )}
            </button>
          )}
        </div>

        {expanded && (
          <div className="process-list">
            {visibleProcessCount === 0 ? (
              <div className="empty">no shells</div>
            ) : (
              <>
                {ptySessions.map((session) => (
                  <PtySessionListItem
                    key={session.id}
                    session={session}
                    active={ptyId === session.id}
                    onSelect={() => selectPtySession(session.id)}
                    onClose={() => closePtySession(session.id)}
                  />
                ))}
                {visibleProcesses.map((proc) => (
                  <ProcessListItem
                    key={proc.id}
                    proc={proc}
                    active={selectedId === proc.id}
                    onSelect={() => {
                      setSelectedId(proc.id);
                      setPtyId(null);
                      setActiveVmId(null);
                      setBrowseUrl(null);
                      setShowSettings(false);
                      setShowHelp(false);
                    }}
                    onKill={() => kill(proc.id)}
                    onDelete={() => deleteProcess(proc.id)}
                    onArchive={() => archiveProcess(proc.id)}
                    onRerun={() => rerun(proc.command)}
                  />
                ))}
              </>
            )}
            {browseHist.length > 0 && (
              <div className="history-section">
                <div className="sidebar-title hist-title">history ({browseHist.length})</div>
                {[...browseHist]
                  .map((u, idx) => ({ u, idx }))
                  .reverse()
                  .map(({ u, idx }) => (
                    <div
                      key={`${idx}-${u}`}
                      className={`history-item ${idx === browseHistIdx ? 'active' : ''}`}
                      title={u}
                      onClick={() => {
                        setBrowseHistIdx(idx);
                        navigate(u);
                      }}
                    >
                      {(() => {
                        try {
                          const p = new URL(u);
                          return p.host + (p.pathname !== '/' ? p.pathname : '');
                        } catch {
                          return u;
                        }
                      })()}
                    </div>
                  ))}
              </div>
            )}
          </div>
        )}
      </div>
      )}

      <div className="pane right">
        <div className="main-output">
          {vmSessions.length > 0 && (
            <div className="vm-tabbar">
              <button
                className={`vm-tab ${activeVmId === null ? 'active' : ''}`}
                onClick={() => setActiveVmId(null)}
                title="shell"
              >
                ⌂ shell
              </button>
              {vmSessions.map((v) => (
                <span
                  key={v.id}
                  className={`vm-tab ${activeVmId === v.id ? 'active' : ''}`}
                >
                  <button
                    className="vm-tab-label"
                    onClick={() => switchVm(v.id)}
                    title={v.label}
                  >
                    🖥 {v.label}
                  </button>
                  <button
                    className="vm-tab-x"
                    title="shut down VM"
                    onClick={() => closeVm(v.id)}
                  >
                    ×
                  </button>
                </span>
              ))}
            </div>
          )}
          {activeVmId !== null ? (
            // A booted VM — its console is a live PTY rendered in xterm.js. The
            // session keeps running in the background when another tab is shown.
            <PtyTerminal
              key={activeVmId}
              ptyId={activeVmId}
              command={vmSessions.find((v) => v.id === activeVmId)?.label || 'vm'}
              onClose={() => closeVm(activeVmId)}
            />
          ) : showHelp ? (
            <HelpOverlay onClose={() => setShowHelp(false)} />
          ) : showSettings ? (
            <SettingsPanel
              onClose={() => setShowSettings(false)}
              session={session}
              onSession={setSession}
              vimEnabled={vimEnabled}
              onVim={setVim}
              agentGenTokens={agentGenTokens}
              onAgentGenTokens={setAgentGenTokens}
              accentColor={accentColor}
              onAccentColor={setAccentColor}
            />
          ) : browseUrl ? (
            // The address lives in the prompt (collapsed URL+REPL line), so this
            // strip is just nav controls — back / forward / reload / close.
            <div className="browser-pane">
              {!browseLoading && browseText && (
                <div className="browser-status">
                  {browseCached ? (
                    <span className="cache-badge cached">⚡ cached</span>
                  ) : (
                    <span className="cache-badge live">● live</span>
                  )}
                  <span className="browse-ms">{browseMs}ms</span>
                </div>
              )}
              {browseLoading ? (
                <AstrolabeSpinner label={browseUrl || 'loading'} />
              ) : (
                <pre className="browser-text">
                  {browseText
                    ? renderRich(browseText, (u) => {
                        setBrowseHist((h) => [...h.slice(0, browseHistIdx + 1), u]);
                        setBrowseHistIdx((i) => i + 1);
                        navigate(u);
                      })
                    : '(no text)'}
                </pre>
              )}
              {browseLinks.length > 0 && (
                <div className="browser-links">
                  <div className="browser-links-head">links ({browseLinks.length})</div>
                  {browseLinks.slice(0, 60).map((l, i) => (
                    <span
                      key={i}
                      className="browser-link"
                      title={l.href}
                      onClick={() => {
                        setBrowseHist((h) => [...h.slice(0, browseHistIdx + 1), l.href]);
                        setBrowseHistIdx((idx) => idx + 1);
                        navigate(l.href);
                      }}
                    >
                      {l.text.slice(0, 60) || l.href}
                    </span>
                  ))}
                </div>
              )}
            </div>
          ) : activePty ? (
            <PtyTerminal
              ptyId={activePty.id}
              command={activePty.command}
              onClose={() => closePtySession(activePty.id)}
              onExit={() => {
                // When the program exits, hand keyboard focus back to the
                // prompt so the next command can be typed immediately (xterm's
                // textarea keeps focus otherwise, swallowing keystrokes), and
                // drop the program name from the window title (back to the cwd).
                markPtyExited(activePty.id);
                setTimeout(() => inputRef.current?.focus(), 0);
              }}
            />
          ) : selectedId ? (
            <ProcessPane processId={selectedId} onNav={openFromOutput} />
          ) : easterEgg ? (
            // Easter egg (double-Shift): e.e. cummings, l(a leaf falls)oneliness.
            <div className="empty-pane" onClick={() => setEasterEgg(false)}>
              <pre className="cummings">{`l(a

le
af
fa

ll

s)
one
l

iness`}</pre>
            </div>
          ) : (
            <div className="empty-pane">run a command, or type a url</div>
          )}
        </div>

        {/* Unified task/command panel — always at the bottom. Holds (when
            relevant) the sidebar-restore button, browser nav controls, the
            prompt (which doubles as the URL bar), and the run button. */}
        <div className="command-box">
          {cwdLabel && !browseUrl && (
            <div className="cwd-bar" title={cwd}>
              {cwdLabel}
              {branch && <span className="cwd-branch" title={`git branch: ${branch}`}> ⎇ {branch}</span>}
            </div>
          )}
          {!browseUrl && (
            <div className="modality-deck" aria-label="agentic modalities">
              {AGENTIC_MODALITIES.map((modality) => (
                <button
                  key={modality.id}
                  type="button"
                  className="modality-button"
                  title={modality.title}
                  disabled={loading}
                  onClick={() => runAgenticModality(modality.command)}
                >
                  {modality.label}
                </button>
              ))}
            </div>
          )}
          <form onSubmit={spawn}>
            {/* Restore the sidebar when it's fully hidden. */}
            {sidebarMode === 'hidden' && (
              <button
                type="button"
                className="btn btn-mini"
                onClick={cycleSidebar}
                title="show sidebar (Cmd/Ctrl-B)"
              >
                ☰
              </button>
            )}
            {/* Browser nav controls appear only while browsing. */}
            {browseUrl && (
              <>
                <button
                  type="button"
                  className="btn btn-mini"
                  onClick={browseBack}
                  disabled={browseHistIdx <= 0}
                  title="back"
                >
                  ‹
                </button>
                <button
                  type="button"
                  className="btn btn-mini"
                  onClick={browseForward}
                  disabled={browseHistIdx >= browseHist.length - 1}
                  title="forward"
                >
                  ›
                </button>
                <button
                  type="button"
                  className="btn btn-mini"
                  onClick={() => navigate(browseUrl, true)}
                  title="reload (bypass cache)"
                >
                  ↻
                </button>
                <button
                  type="button"
                  className="btn btn-mini"
                  onClick={() => { setBrowseUrl(null); setCommand(''); }}
                  title="close page"
                >
                  ✕
                </button>
              </>
            )}
            {vimEnabled && (
              <span className={`vim-badge ${vimMode}`} title="vim mode (toggle in settings)">
                {vimMode === 'normal' ? 'NOR' : 'INS'}
              </span>
            )}
            {searchMode && (
              <span
                className={`search-badge ${searchNoMatch ? 'nomatch' : ''}`}
                title="reverse history search (Ctrl-R again = older, Enter = run, Esc = cancel)"
              >
                {searchNoMatch ? 'failing ' : ''}(reverse-i-search)`{searchQuery}`
              </span>
            )}
            <textarea
              ref={inputRef}
              rows={command.includes('\n') ? Math.min(command.split('\n').length, 12) : 1}
              placeholder={browseUrl ? 'url or command' : STARTER_COMMAND}
              value={command}
              onChange={(e) => setCommand(e.target.value)}
              onKeyDown={onKeyDown}
              disabled={loading}
              className="input input-multiline"
              autoFocus
              spellCheck={false}
              autoCorrect="off"
              autoCapitalize="off"
              autoComplete="off"
            />
            <button type="submit" disabled={loading} className="btn btn-primary">
              {loading ? '...' : browseUrl ? 'go' : 'run'}
            </button>
          </form>
          {completions.length > 1 && (
            <div className="completions">
              {completions.slice(0, 40).map((c, i) => (
                <span
                  key={c}
                  className={`completion ${i === completeIdx ? 'active' : ''}`}
                  onClick={() => {
                    setCompleteIdx(i);
                    replaceToken(c);
                    setCompletions([]);
                    inputRef.current?.focus();
                  }}
                >
                  {c}
                </span>
              ))}
              {completions.length > 40 && (
                <span className="completion dim">+{completions.length - 40} more</span>
              )}
            </div>
          )}
          {error && <div className="error-msg">{error}</div>}
        </div>
      </div>
    </div>
  );
}

function ProcessPane({
  processId,
  onNav,
}: {
  processId: string;
  onNav?: (url: string) => void;
}) {
  const [proc, setProc] = useState<ProcessInfo | null>(null);
  const outputRef = useRef<HTMLPreElement>(null);
  const findInputRef = useRef<HTMLInputElement>(null);
  // True once the user scrolls away from the bottom — suppresses tail-follow so
  // streaming output doesn't yank them back while they read.
  const userScrolledUpRef = useRef(false);
  const onOutputScroll = () => {
    const pre = outputRef.current;
    if (pre) userScrolledUpRef.current = pre.scrollHeight - pre.scrollTop - pre.clientHeight >= 60;
  };
  // Find-in-output (Cmd-F): a small find bar highlights matches in the output
  // and jumps between them.
  const [findOpen, setFindOpen] = useState(false);
  const [findQuery, setFindQuery] = useState('');
  const [findCurrent, setFindCurrent] = useState(0);

  // Cmd/Ctrl-F (without the Ctrl-fullscreen combo) opens the find bar for the
  // shown output. PageUp/PageDown (and Cmd/Ctrl-Home/End) scroll the output
  // pane from the keyboard — the <pre> isn't focusable, so without this you
  // can't review long output without the mouse. Registered while mounted.
  useEffect(() => {
    const onKey = (e: KeyboardEvent) => {
      if ((e.metaKey || e.ctrlKey) && !e.altKey && !(e.metaKey && e.ctrlKey) && e.key.toLowerCase() === 'f') {
        e.preventDefault();
        setFindOpen(true);
        setTimeout(() => findInputRef.current?.focus(), 0);
        return;
      }
      const pre = outputRef.current;
      if (!pre) return;
      const page = Math.max(40, pre.clientHeight * 0.9);
      if (e.key === 'PageUp') {
        e.preventDefault();
        pre.scrollBy({ top: -page });
      } else if (e.key === 'PageDown') {
        e.preventDefault();
        pre.scrollBy({ top: page });
      } else if ((e.metaKey || e.ctrlKey) && e.key === 'Home') {
        e.preventDefault();
        pre.scrollTo({ top: 0 });
      } else if ((e.metaKey || e.ctrlKey) && e.key === 'End') {
        e.preventDefault();
        pre.scrollTo({ top: pre.scrollHeight });
      }
    };
    window.addEventListener('keydown', onKey);
    return () => window.removeEventListener('keydown', onKey);
  }, []);

  // Total matches in the (ANSI-stripped) output for the current query.
  const findTotal = (() => {
    if (!findOpen || !findQuery) return 0;
    const hay = stripAnsi(proc?.output || '').toLowerCase();
    const q = findQuery.toLowerCase();
    let n = 0;
    let i = hay.indexOf(q);
    while (i !== -1) {
      n++;
      i = hay.indexOf(q, i + q.length);
    }
    return n;
  })();

  // Keep the current-match index in range, and scroll the active hit into view.
  useEffect(() => {
    if (findTotal > 0 && findCurrent >= findTotal) setFindCurrent(0);
    const el = outputRef.current?.querySelector('.find-hit.current');
    if (el) (el as HTMLElement).scrollIntoView({ block: 'center', behavior: 'smooth' });
  }, [findCurrent, findQuery, findTotal, proc?.output]);

  const findStep = (dir: 1 | -1) => {
    if (findTotal === 0) return;
    setFindCurrent((c) => (c + dir + findTotal) % findTotal);
  };
  const closeFind = () => {
    setFindOpen(false);
    setFindQuery('');
    setFindCurrent(0);
  };

  // The backend's get_process is the single source of truth (full output + exit
  // code). We refresh it on a timer AND on each live shell:output event — both
  // paths just re-read the authoritative buffer, so output can never double up.
  // `refreshing` coalesces bursts: while one read is in flight, further triggers
  // are dropped (the in-flight read already returns the latest buffer).
  useEffect(() => {
    let cancelled = false;
    let timer: ReturnType<typeof setTimeout> | null = null;
    let unsubscribe: (() => void) | null = null;
    let refreshing = false;

    const refreshNow = async () => {
      if (cancelled || refreshing) return;
      refreshing = true;
      try {
        const p: ProcessInfo = await invoke('get_process', { processId });
        if (!cancelled) setProc(p);
      } catch {
        /* transient; the timer retries */
      } finally {
        refreshing = false;
      }
    };

    // Timer: source-of-truth reconciliation + completion detection. Fast while
    // running (recovers fast commands that finish before mount), slow once done.
    let doneTicks = 0;
    const tick = async () => {
      if (cancelled) return;
      try {
        const p: ProcessInfo = await invoke('get_process', { processId });
        if (cancelled) return;
        setProc(p);
        if (p.status !== 'running') {
          // Done: poll a couple more times to catch the final buffered flush,
          // then STOP. The output can't change once finished, so continuing to
          // re-poll (and re-parse a large output every second) is wasted work.
          if (++doneTicks > 2) return;
          timer = setTimeout(tick, 400);
        } else {
          doneTicks = 0;
          timer = setTimeout(tick, POLL_MS);
        }
      } catch {
        if (!cancelled) timer = setTimeout(tick, 1000);
      }
    };
    tick();

    // Live fast-path: each emitted line triggers an immediate authoritative
    // refresh (not a blind append), so streaming output appears with ~0 latency
    // instead of waiting up to POLL_MS — and stays duplication-proof.
    (async () => {
      try {
        unsubscribe = await listen<{ process_id: string; line: string }>(
          'shell:output',
          (event) => {
            if (event.payload.process_id === processId) refreshNow();
          }
        );
      } catch (e) {
        console.error('Failed to set up event listener:', e);
      }
    })();

    return () => {
      cancelled = true;
      if (timer) clearTimeout(timer);
      if (unsubscribe) unsubscribe();
    };
  }, [processId]);

  // Follow the tail as output streams in — unless the user has scrolled up to
  // read (tracked by onOutputScroll), in which case don't yank them down.
  useEffect(() => {
    const pre = outputRef.current;
    if (!pre) return;
    if (!userScrolledUpRef.current) pre.scrollTop = pre.scrollHeight;
  }, [proc?.output]);

  if (!proc) {
    return <div className="empty-pane">loading...</div>;
  }

  const hasOutput = proc.output && proc.output.length > 0;
  const running = proc.status === 'running';
  const ok = proc.exit_code === 0;

  const copyOutput = () => {
    navigator.clipboard?.writeText(stripAnsi(proc.output || '')).catch(() => {});
  };

  return (
    <div className="output-pane">
      <div className="output-header">
        <span className="output-cmd">{proc.command}</span>
        {running && (
          <button
            className="btn btn-mini stop"
            onClick={() => invoke('interrupt_command').catch(() => {})}
            title="interrupt — restarts the REPL (Ctrl-C)"
          >
            stop
          </button>
        )}
        {hasOutput && (
          <button
            className={`btn btn-mini ${findOpen ? 'active' : ''}`}
            onClick={() => {
              setFindOpen((o) => !o);
              setTimeout(() => findInputRef.current?.focus(), 0);
            }}
            title="find in output (Cmd/Ctrl-F)"
          >
            find
          </button>
        )}
        {hasOutput && (
          <button className="btn btn-mini" onClick={copyOutput} title="copy output">
            copy
          </button>
        )}
      </div>
      {findOpen && (
        <div className="find-bar">
          <input
            ref={findInputRef}
            className="find-input"
            placeholder="find in output…"
            value={findQuery}
            spellCheck={false}
            onChange={(e) => {
              setFindQuery(e.target.value);
              setFindCurrent(0);
            }}
            onKeyDown={(e) => {
              if (e.key === 'Enter') {
                e.preventDefault();
                findStep(e.shiftKey ? -1 : 1);
              } else if (e.key === 'Escape') {
                e.preventDefault();
                closeFind();
              }
            }}
          />
          <span className="find-count">
            {findQuery ? `${findTotal === 0 ? 0 : findCurrent + 1}/${findTotal}` : ''}
          </span>
          <button className="btn btn-mini" onClick={() => findStep(-1)} title="previous (Shift-Enter)">
            ↑
          </button>
          <button className="btn btn-mini" onClick={() => findStep(1)} title="next (Enter)">
            ↓
          </button>
          <button className="btn btn-mini" onClick={closeFind} title="close (Esc)">
            ✕
          </button>
        </div>
      )}
      <pre className="output-text" ref={outputRef} onScroll={onOutputScroll}>
        {hasOutput
          ? renderAnsi(
              proc.output,
              onNav,
              findOpen && findQuery ? { query: findQuery, current: findCurrent } : undefined
            )
          : running
          ? '(running...)'
          : '(no output)'}
      </pre>
      <div className="status-bar">
        {running ? (
          <span className="run-dot">● running</span>
        ) : proc.exit_code !== null ? (
          <span className={`exit-pill ${ok ? 'ok' : 'bad'}`}>
            {ok ? '✓' : '✗'} exit {proc.exit_code}
          </span>
        ) : (
          <span className="dim">done</span>
        )}
        {proc.duration_ms !== null && proc.duration_ms !== undefined && (
          <span className="cmd-ms">
            {proc.duration_ms < 1 ? '<1ms' : `${proc.duration_ms}ms`}
          </span>
        )}
      </div>
    </div>
  );
}

/// Keyboard-shortcut cheat sheet (Cmd/Ctrl-/ or the ? button). Surfaces the
/// terminal's full hotkey set, which had grown rich but undiscoverable.
function HelpOverlay({ onClose }: { onClose: () => void }) {
  const isMac =
    typeof navigator !== 'undefined' && /Mac|iPhone|iPad/.test(navigator.platform);
  const mod = isMac ? '⌘' : 'Ctrl';
  const groups: { title: string; keys: [string, string][] }[] = [
    {
      title: 'general',
      keys: [
        [`${mod}-B`, 'cycle sidebar: rail / expanded / hidden'],
        [`${mod}-,`, 'settings · ~/.moonshine'],
        [`${mod}-/`, 'this shortcut sheet'],
        [`${mod}-K`, 'clear finished command history'],
        [`${mod}-Enter`, 'focus the prompt'],
        [`${mod}-Ctrl-F / F11`, 'toggle fullscreen'],
        [`${mod}-+ / ${mod}-- / ${mod}-0`, 'zoom in / out / reset'],
        ['Esc', 'close an overlay'],
        ['Shift Shift', 'a small surprise'],
      ],
    },
    {
      title: 'prompt · history',
      keys: [
        ['Enter', 'run command'],
        ['Shift-Enter', 'insert a newline (multi-line)'],
        ['Up / Down', 'previous / next command'],
        ['Ctrl-R', 'reverse-incremental history search'],
        [`${mod}-[ / ${mod}-]`, 'review older / newer command output'],
        ['Tab', 'complete command / path'],
      ],
    },
    {
      title: 'prompt · readline editing',
      keys: [
        ['Ctrl-A / Ctrl-E', 'jump to line start / end'],
        ['Ctrl-K / Ctrl-U', 'kill to end / to start'],
        ['Ctrl-W', 'delete the word before the caret'],
        ['Ctrl-L', 'clear the output pane'],
        ['Ctrl-C', 'interrupt a running command, else clear the line'],
      ],
    },
    {
      title: 'output',
      keys: [
        ['PageUp / PageDown', 'scroll the output'],
        [`${mod}-Home / ${mod}-End`, 'jump to top / bottom'],
        [`${mod}-F`, 'find in output'],
      ],
    },
    {
      title: 'running programs',
      keys: [
        ['Ctrl-C / stop', 'interrupt a wedged command (restarts the REPL)'],
        ['a url / domain', 'opens the sovereign browser pane'],
        ['vim · claude · python', 'run in a real PTY (full-screen TUI via xterm)'],
      ],
    },
  ];
  return (
    <div className="settings-panel help-panel">
      <div className="settings-head">
        <span className="settings-title">keyboard shortcuts</span>
        <button className="btn btn-toggle" onClick={onClose} title="close (Esc)">
          ✕
        </button>
      </div>
      <div className="settings-body">
        {groups.map((g) => (
          <div key={g.title} className="help-group">
            <div className="settings-section">{g.title}</div>
            {g.keys.map(([k, desc]) => (
              <div key={k} className="help-row">
                <kbd className="help-key">{k}</kbd>
                <span className="help-desc">{desc}</span>
              </div>
            ))}
          </div>
        ))}
        <div className="settings-hint">
          vim mode (toggle in settings) adds modal motions; readline keys still
          work in insert mode.
        </div>
      </div>
    </div>
  );
}

function formatBytes(n: number | null): string {
  if (n === null) return '';
  const units = ['B', 'KB', 'MB', 'GB', 'TB'];
  let value = n;
  let unit = 0;
  while (value >= 1024 && unit < units.length - 1) {
    value /= 1024;
    unit += 1;
  }
  return `${value >= 10 || unit === 0 ? value.toFixed(0) : value.toFixed(1)} ${units[unit]}`;
}

function ModelManager() {
  const [status, setStatus] = useState<ModelStatus | null>(null);
  const [selected, setSelected] = useState('');
  const [sourceUrl, setSourceUrl] = useState('');
  const [busy, setBusy] = useState('');
  const [err, setErr] = useState('');

  const refresh = () =>
    invoke<ModelStatus>('model_status')
      .then((next) => {
        setStatus(next);
        const active = next.models.find((model) => model.active) || next.models[0];
        if (active) {
          setSelected(active.id);
          setSourceUrl(active.url);
        }
        setErr('');
      })
      .catch((e) => setErr(String(e)));

  useEffect(() => {
    refresh();
  }, []);

  const active = status?.models.find((model) => model.id === selected) || status?.models[0];
  const choose = (modelId: string) => {
    const model = status?.models.find((m) => m.id === modelId);
    setSelected(modelId);
    setSourceUrl(model?.url || '');
  };
  const configure = () => {
    if (!selected) return;
    setBusy('configure');
    setErr('');
    invoke<ModelSnapshot>('configure_model', { modelId: selected, url: sourceUrl })
      .then(() => refresh())
      .catch((e) => setErr(String(e)))
      .finally(() => setBusy(''));
  };
  const download = () => {
    if (!selected) return;
    setBusy('download');
    setErr('');
    invoke<ModelSnapshot>('download_model', { modelId: selected })
      .then(() => refresh())
      .catch((e) => setErr(String(e)))
      .finally(() => setBusy(''));
  };

  return (
    <>
      <div className="settings-section">local model</div>
      {!status && !err && <div className="dim sm">loading…</div>}
      {err && <div className="error-msg">{err}</div>}
      {status && active && (
        <div className="model-manager">
          <div className="settings-row">
            <span className="settings-label">active</span>
            <span className="settings-value model-controls">
              <select
                className="input-sm model-select"
                value={selected}
                onChange={(e) => choose(e.target.value)}
              >
                {status.models.map((model) => (
                  <option key={model.id} value={model.id}>
                    {model.label}
                  </option>
                ))}
              </select>
              <button
                className="btn btn-mini"
                onClick={configure}
                disabled={!!busy}
              >
                set
              </button>
            </span>
          </div>
          <div className="settings-row">
            <span className="settings-label">cache</span>
            <span className="settings-value">
              {active.present ? (
                <span className="ok">cached {formatBytes(active.size_bytes)}</span>
              ) : (
                <span className="dim">missing</span>
              )}
            </span>
          </div>
          <div className="settings-row">
            <span className="settings-label">path</span>
            <span className="settings-value">
              <code>{active.cache_path}</code>
            </span>
          </div>
          <div className="settings-row">
            <span className="settings-label">source</span>
            <span className="settings-value model-source">
              <input
                className="input-sm"
                value={sourceUrl}
                onChange={(e) => setSourceUrl(e.target.value)}
                spellCheck={false}
              />
              <button
                className="btn btn-mini btn-primary"
                onClick={download}
                disabled={!!busy}
              >
                {busy === 'download' ? 'downloading…' : 'download'}
              </button>
            </span>
          </div>
          <div className="settings-row">
            <span className="settings-label">cache dir</span>
            <span className="settings-value">
              <code>{status.cache_dir}</code>
            </span>
          </div>
        </div>
      )}
    </>
  );
}

function SkymeshManager() {
  const [status, setStatus] = useState<SkymeshStatus | null>(null);
  const [busy, setBusy] = useState(false);
  const [err, setErr] = useState('');

  const refresh = () =>
    invoke<SkymeshStatus>('skymesh_status')
      .then((next) => {
        setStatus(next);
        setErr('');
      })
      .catch((e) => setErr(String(e)));

  useEffect(() => {
    refresh();
  }, []);

  const toggle = () => {
    if (!status) return;
    setBusy(true);
    setErr('');
    invoke<SkymeshStatus>('configure_skymesh', { enabled: !status.enabled })
      .then(setStatus)
      .catch((e) => setErr(String(e)))
      .finally(() => setBusy(false));
  };

  return (
    <>
      <div className="settings-section">skymesh</div>
      {!status && !err && <div className="dim sm">loading…</div>}
      {err && <div className="error-msg">{err}</div>}
      {status && (
        <div className="model-manager">
          <div className="settings-toggle">
            <span>
              protocol69 mesh client
              <div className="dim sm">joins global + local skymesh and fills cache keys</div>
            </span>
            <button
              className={`toggle-switch ${status.enabled ? 'on' : ''}`}
              onClick={toggle}
              disabled={busy}
            >
              {status.enabled ? 'on' : 'off'}
            </button>
          </div>
          <div className="settings-row">
            <span className="settings-label">bridge</span>
            <span className="settings-value">
              {status.running ? (
                <span className="ok">running{status.pid ? ` · pid ${status.pid}` : ''}</span>
              ) : status.enabled ? (
                <span className="dim">starting</span>
              ) : (
                <span className="dim">disabled</span>
              )}
            </span>
          </div>
          {status.node && (
            <div className="settings-row">
              <span className="settings-label">node</span>
              <span className="settings-value">
                <code>{status.node}</code>
              </span>
            </div>
          )}
          {status.model && (
            <div className="settings-row">
              <span className="settings-label">model</span>
              <span className="settings-value">
                <code>{status.model}</code>
              </span>
            </div>
          )}
          {status.last_cache_key && (
            <div className="settings-row">
              <span className="settings-label">last key</span>
              <span className="settings-value">
                <code>{status.last_cache_key}</code>
              </span>
            </div>
          )}
          {status.filled_cache_keys.length > 0 && (
            <div className="settings-row">
              <span className="settings-label">filled keys</span>
              <span className="settings-value key-list">
                {status.filled_cache_keys.map((key) => (
                  <code key={key}>{key}</code>
                ))}
              </span>
            </div>
          )}
          {status.last_error && <div className="error-msg">{status.last_error}</div>}
        </div>
      )}
    </>
  );
}

function MobileBridgeManager() {
  const [status, setStatus] = useState<MobileBridgeStatus | null>(null);
  const [url, setUrl] = useState('');
  const [token, setToken] = useState('');
  const [busy, setBusy] = useState(false);
  const [err, setErr] = useState('');

  const refresh = () =>
    invoke<MobileBridgeStatus>('mobile_bridge_status')
      .then((next) => {
        setStatus(next);
        setUrl(next.url);
        setToken('');
        setErr('');
      })
      .catch((e) => setErr(String(e)));

  useEffect(() => {
    refresh();
  }, []);

  const save = (enabled = status?.enabled ?? false) => {
    setBusy(true);
    setErr('');
    invoke<MobileBridgeStatus>('configure_mobile_bridge', {
      enabled,
      url,
      token: token.trim() ? token : null,
    })
      .then((next) => {
        setStatus(next);
        setUrl(next.url);
        setToken('');
      })
      .catch((e) => setErr(String(e)))
      .finally(() => setBusy(false));
  };

  return (
    <>
      <div className="settings-section">mobile bridge</div>
      {!status && !err && <div className="dim sm">loading…</div>}
      {err && <div className="error-msg">{err}</div>}
      {status && (
        <div className="model-manager">
          <div className="settings-toggle">
            <span>
              network shell bridge
              <div className="dim sm">optional LAN handoff for iOS commands</div>
            </span>
            <button
              className={`toggle-switch ${status.enabled ? 'on' : ''}`}
              onClick={() => save(!status.enabled)}
              disabled={busy}
            >
              {status.enabled ? 'on' : 'off'}
            </button>
          </div>
          <div className="settings-row">
            <span className="settings-label">URL</span>
            <span className="settings-value model-source">
              <input
                className="input-sm"
                value={url}
                onChange={(e) => setUrl(e.target.value)}
                placeholder="http://10.0.0.12:14231"
                spellCheck={false}
              />
              <button className="btn btn-mini" onClick={() => save()} disabled={busy}>
                save
              </button>
            </span>
          </div>
          <div className="settings-row">
            <span className="settings-label">token</span>
            <span className="settings-value model-source">
              <input
                className="input-sm"
                value={token}
                onChange={(e) => setToken(e.target.value)}
                placeholder={status.token_configured ? 'configured' : 'optional'}
                spellCheck={false}
              />
            </span>
          </div>
          <div className="settings-row">
            <span className="settings-label">config</span>
            <span className="settings-value">
              <code>{status.config_path}</code>
            </span>
          </div>
          {status.last_error && <div className="error-msg">{status.last_error}</div>}
        </div>
      )}
    </>
  );
}

/// Read-only settings + ~/.moonshine status panel. Surfaces the resolved
/// binaries, guard home, identity markers, and live REPL state so the operator
/// can confirm the sovereign toolchain is wired correctly.
function SettingsPanel({
  onClose,
  session,
  onSession,
  vimEnabled,
  onVim,
  agentGenTokens,
  onAgentGenTokens,
  accentColor,
  onAccentColor,
}: {
  onClose: () => void;
  session: AuthSession | null;
  onSession: (s: AuthSession | null) => void;
  vimEnabled: boolean;
  onVim: (on: boolean) => void;
  agentGenTokens: number;
  onAgentGenTokens: (tokens: number) => void;
  accentColor: string;
  onAccentColor: (color: string) => void;
}) {
  const [info, setInfo] = useState<SystemInfo | null>(null);
  const [err, setErr] = useState('');
  const [signinErr, setSigninErr] = useState('');
  const [signingIn, setSigningIn] = useState(false);
  const [accentDraft, setAccentDraft] = useState(accentColor);

  useEffect(() => {
    setAccentDraft(accentColor);
  }, [accentColor]);

  const commitAccentDraft = () => {
    const normalized = normalizeAccentColor(accentDraft);
    onAccentColor(normalized);
    setAccentDraft(normalized);
  };

  const signIn = async (provider: string) => {
    setSigningIn(true);
    setSigninErr('');
    try {
      onSession(await runCardYogaSignin(provider));
    } catch (e) {
      setSigninErr(String(e));
    } finally {
      setSigningIn(false);
    }
  };
  const signOut = () => {
    localStorage.removeItem(AUTH_STORAGE_KEY);
    onSession(null);
  };

  useEffect(() => {
    invoke<SystemInfo>('system_info')
      .then(setInfo)
      .catch((e) => setErr(String(e)));
  }, []);

  const row = (label: string, value: React.ReactNode) => (
    <div className="settings-row">
      <span className="settings-label">{label}</span>
      <span className="settings-value">{value}</span>
    </div>
  );

  return (
    <div className="settings-panel">
      <div className="settings-head">
        <span className="settings-title">settings · ~/.moonshine</span>
        <button className="btn btn-toggle" onClick={onClose} title="close">
          ✕
        </button>
      </div>
      {err && <div className="error-msg">{err}</div>}
      {!info && !err && <div className="empty-pane">loading…</div>}
      {info && (
        <div className="settings-body">
          <div className="settings-section">runtime</div>
          <div className={`doctor-card ${info.ready ? 'ready' : 'blocked'}`}>
            <div className="doctor-head">
              <span className={`status-dot ${info.ready ? 'ok' : 'bad'}`} />
              <strong>{info.ready ? 'ready' : 'needs attention'}</strong>
            </div>
            {info.problems.length > 0 ? (
              <ul className="doctor-list">
                {info.problems.map((problem) => (
                  <li key={problem}>{problem}</li>
                ))}
              </ul>
            ) : (
              <span className="dim sm">runtime binaries, identity, and guard home are available</span>
            )}
          </div>

          <ModelManager />
          <SkymeshManager />
          <MobileBridgeManager />

          <div className="settings-section">appearance</div>
          <div className="settings-row">
            <span className="settings-label">accent</span>
            <span className="settings-value accent-color-control">
              <input
                className="accent-color-picker"
                type="color"
                value={accentColor}
                onChange={(e) => onAccentColor(e.currentTarget.value)}
                title="pick accent color"
              />
              <input
                className="input-sm accent-color-input"
                value={accentDraft}
                onChange={(e) => setAccentDraft(e.currentTarget.value)}
                onBlur={commitAccentDraft}
                onKeyDown={(e) => {
                  if (e.key === 'Enter') {
                    e.preventDefault();
                    commitAccentDraft();
                  }
                }}
                aria-label="accent hex color"
              />
              {ACCENT_SWATCHES.map((color) => (
                <button
                  key={color}
                  className={`accent-swatch ${accentColor === color ? 'active' : ''}`}
                  style={{ backgroundColor: color }}
                  onClick={() => onAccentColor(color)}
                  title={color}
                />
              ))}
              <button
                className="btn btn-mini"
                onClick={() => onAccentColor(DEFAULT_ACCENT_COLOR)}
                title="reset accent color"
              >
                reset
              </button>
            </span>
          </div>

          <div className="settings-section">agent responses</div>
          <div className="settings-row">
            <span className="settings-label">token cap</span>
            <span className="settings-value token-cap-control">
              <input
                className="input-sm token-cap-input"
                type="number"
                min={MIN_AGENT_GEN_TOKENS}
                max={MAX_AGENT_GEN_TOKENS}
                step={1}
                value={agentGenTokens}
                onChange={(e) => onAgentGenTokens(Number(e.currentTarget.value))}
                title="MOONSHINE_AGENT_GEN_TOKENS"
              />
              {[128, 256, 512].map((tokens) => (
                <button
                  key={tokens}
                  className={`cap-chip ${agentGenTokens === tokens ? 'active' : ''}`}
                  onClick={() => onAgentGenTokens(tokens)}
                  title={`set response cap to ${tokens} generated tokens`}
                >
                  {tokens}
                </button>
              ))}
            </span>
          </div>
          <div className="settings-hint compact">
            Applies to shell fallback and <code>agent chat</code> turns through{' '}
            <code>MOONSHINE_AGENT_GEN_TOKENS</code>. A changed cap restarts the
            warm REPL before the next command.
          </div>

          <div className="settings-section">account · card.yoga</div>
          {session ? (
            <div className="account-card">
              {session.identity.picture ? (
                <img className="account-avatar" src={session.identity.picture} alt="" />
              ) : (
                <span className="account-avatar avatar-initial">
                  {(session.identity.name || session.identity.email || '?')
                    .charAt(0)
                    .toUpperCase()}
                </span>
              )}
              <div className="account-meta">
                <div className="account-name">
                  {session.identity.name || session.identity.email || 'signed in'}
                </div>
                <code className="account-did">{session.rootDid}</code>
              </div>
              <button className="btn btn-mini" onClick={signOut}>
                sign out
              </button>
            </div>
          ) : (
            <div className="signin-row">
              <span className="dim">sign in for a DID-token session</span>
              <div className="signin-buttons">
                {['google', 'github', 'apple'].map((p) => (
                  <button
                    key={p}
                    className="btn btn-mini"
                    onClick={() => signIn(p)}
                    disabled={signingIn}
                  >
                    {p}
                  </button>
                ))}
              </div>
              {signingIn && <span className="dim sm">opening browser… complete sign-in there</span>}
              {signinErr && <div className="error-msg">{signinErr}</div>}
            </div>
          )}

          <div className="settings-section">identity</div>
          {row(
            'status',
            info.identity_present ? (
              <span className="ok">admitted</span>
            ) : (
              <span className="bad">no identity — run `moonshine init`</span>
            )
          )}
          {info.identity_path && row('path', <code>{info.identity_path}</code>)}
          {row('REPL', info.repl_active ? (
            <span className="ok">live (sub-1ms)</span>
          ) : (
            <span className="dim">idle — starts on first command</span>
          ))}

          <div className="settings-section">sovereign toolchain</div>
          {row('moonshine', <code>{info.moonshine_bin}</code>)}
          {row('monster-guard', <code>{info.guard_bin}</code>)}
          {row('guard home', <code>{info.guard_home}</code>)}

          <AdmissionsManager />

          <div className="settings-section">editing</div>
          <div className="settings-toggle">
            <span>
              vim keybindings
              <div className="dim sm">
                normal/insert modes on the prompt — i/a, h/l, w/b, 0/$, x, D, dd,
                j/k history, Esc to normal
              </div>
            </span>
            <button
              className={`toggle-switch ${vimEnabled ? 'on' : ''}`}
              onClick={() => onVim(!vimEnabled)}
            >
              {vimEnabled ? 'on' : 'off'}
            </button>
          </div>

          <div className="settings-hint">
            Binary paths come from <code>MOONSHINE_BIN</code> /{' '}
            <code>MONSTER_GUARD_BIN</code> (env) then <code>PATH</code>. The guard
            home holds UCAN grants + revocations.
          </div>
        </div>
      )}
    </div>
  );
}

/// Manage admitted binaries and their UCAN capability delegations: list each
/// grant (name, caps, expiry), add a new admission, lock one down (revoke), or
/// remove it entirely. All mutations shell out to monster-guard via the backend.
function AdmissionsManager() {
  const [items, setItems] = useState<Admission[] | null>(null);
  const [err, setErr] = useState('');
  const [busy, setBusy] = useState('');
  const [adding, setAdding] = useState(false);
  const [newName, setNewName] = useState('');
  const [newCaps, setNewCaps] = useState<string[]>(['exec']);
  const [newTtlDays, setNewTtlDays] = useState(30);

  const reload = async () => {
    try {
      setItems(await invoke<Admission[]>('list_admissions'));
      setErr('');
    } catch (e) {
      setErr(String(e));
    }
  };
  useEffect(() => {
    reload();
  }, []);

  const toggleCap = (cap: string) =>
    setNewCaps((cs) => (cs.includes(cap) ? cs.filter((c) => c !== cap) : [...cs, cap]));

  const submitNew = async () => {
    const name = newName.trim();
    if (!name) return;
    setBusy(name);
    try {
      await invoke('delegate_admission', {
        name,
        caps: newCaps,
        ttlSecs: Math.max(1, Math.round(newTtlDays * 86400)),
      });
      setNewName('');
      setNewCaps(['exec']);
      setAdding(false);
      await reload();
    } catch (e) {
      setErr(String(e));
    } finally {
      setBusy('');
    }
  };

  const revoke = async (name: string) => {
    setBusy(name);
    try {
      await invoke('revoke_admission', { name });
      await reload();
    } catch (e) {
      setErr(String(e));
    } finally {
      setBusy('');
    }
  };

  const remove = async (name: string) => {
    setBusy(name);
    try {
      await invoke('remove_admission', { name });
      await reload();
    } catch (e) {
      setErr(String(e));
    } finally {
      setBusy('');
    }
  };

  return (
    <>
      <div className="settings-section">
        admitted binaries ({items?.length ?? 0})
        <button
          className="btn btn-mini"
          onClick={() => setAdding((a) => !a)}
          title="admit a new binary"
        >
          {adding ? '−' : '+ admit'}
        </button>
      </div>

      {adding && (
        <div className="admit-form">
          <input
            className="input input-sm"
            placeholder="binary name (e.g. claude)"
            value={newName}
            onChange={(e) => setNewName(e.target.value)}
            spellCheck={false}
            autoCapitalize="off"
            autoCorrect="off"
          />
          <div className="cap-row">
            {ALL_CAPS.map((c) => (
              <button
                key={c}
                className={`cap-chip ${newCaps.includes(c) ? 'on' : ''}`}
                onClick={() => toggleCap(c)}
                type="button"
              >
                {c}
              </button>
            ))}
            <span className="ttl-field">
              ttl
              <input
                className="input input-xs"
                type="number"
                min={1}
                value={newTtlDays}
                onChange={(e) => setNewTtlDays(Number(e.target.value))}
              />
              d
            </span>
            <button className="btn btn-mini btn-primary" onClick={submitNew} disabled={!!busy}>
              admit
            </button>
          </div>
        </div>
      )}

      {err && <div className="error-msg">{err}</div>}
      {items && items.length === 0 && <div className="empty">no admitted binaries</div>}
      {items &&
        items.map((a) => (
          <div className="admission-row" key={a.name}>
            <div className="adm-main">
              <code className={a.revoked || a.expired ? 'dim' : ''}>{a.name}</code>
              <span className="adm-caps">
                {a.caps.length ? (
                  a.caps.map((c) => (
                    <span className="cap-chip on sm" key={c}>
                      {c}
                    </span>
                  ))
                ) : (
                  <span className="dim">no caps</span>
                )}
              </span>
              {a.revoked ? (
                <span className="bad sm">revoked</span>
              ) : a.expired ? (
                <span className="dim sm">expired</span>
              ) : (
                <span className="ok sm">active</span>
              )}
            </div>
            <div className="adm-actions">
              {!a.revoked && (
                <button
                  className="btn btn-mini"
                  onClick={() => revoke(a.name)}
                  disabled={busy === a.name}
                  title="lock down — revoke this grant"
                >
                  lock
                </button>
              )}
              <button
                className="btn btn-mini btn-danger"
                onClick={() => remove(a.name)}
                disabled={busy === a.name}
                title="remove admission entirely"
              >
                remove
              </button>
            </div>
          </div>
        ))}
    </>
  );
}

/// Live terminal for an interactive PTY session. Streams `pty:output` into a
/// scrollback buffer, forwards keystrokes to the program via `pty_write`, and
/// strips ANSI for a clean (if non-cursor-addressable) view. Good enough for
/// line-oriented interactive tools (claude, python, ssh prompts); full-screen
/// curses apps (vim) will look rough until we add an xterm.js renderer.
// Vendored xterm.js exposes window.Terminal + window.FitAddon (UMD).
type XTerm = {
  open(el: HTMLElement): void;
  write(data: string): void;
  onData(cb: (data: string) => void): void;
  dispose(): void;
  focus(): void;
  loadAddon(addon: unknown): void;
  refresh(start: number, end: number): void;
  clearTextureAtlas?: () => void;
  cols: number;
  rows: number;
};
type XFitAddon = { fit(): void };
declare global {
  interface Window {
    Terminal?: new (opts?: Record<string, unknown>) => XTerm;
    FitAddon?: { FitAddon: new () => XFitAddon };
  }
}

// xterm mounts its DOM into this div imperatively. React must NEVER reconcile
// it, or a parent re-render (cwd polling, etc.) wipes xterm's appended nodes
// (observed: host stays connected but childCount drops to 0, terminal goes
// blank while the Terminal JS object lives on). memo with an always-equal
// comparator freezes the node so React leaves xterm's children alone.
const XtermHost = React.memo(
  React.forwardRef<HTMLDivElement>(function XtermHost(_props, ref) {
    return <div className="xterm-host" ref={ref} />;
  }),
  () => true
);

function PtyTerminal({
  ptyId,
  command,
  onClose,
  onExit,
}: {
  ptyId: string;
  command: string;
  onClose: () => void;
  onExit?: () => void;
}) {
  const [buffer, setBuffer] = useState('');
  const [exited, setExited] = useState(false);
  // Full VT emulation via vendored xterm.js v5.5.0 (+ FitAddon). DEFAULT ON —
  // full-screen TUIs (vim, top, claude) render in a real terminal grid. The
  // long-standing blank render was NOT a renderer bug: the host div is
  // React-owned, so a parent re-render reconciled away xterm's imperatively
  // appended DOM (host stayed connected, childCount → 0). Fixed by mounting
  // into <XtermHost>, a memoized never-re-rendering node. Set moonshine.xterm=0
  // to fall back to the ANSI-stripped scrollback pane.
  const useXterm =
    typeof window !== 'undefined' &&
    !!window.Terminal &&
    !!window.FitAddon &&
    localStorage.getItem('moonshine.xterm') !== '0';
  const hostRef = useRef<HTMLDivElement>(null);
  const screenRef = useRef<HTMLPreElement>(null);
  const termRef = useRef<XTerm | null>(null);

  useEffect(() => {
    let stop = false;
    let unOut: (() => void) | null = null;
    let unExit: (() => void) | null = null;
    let term: XTerm | null = null;
    let fit: XFitAddon | null = null;
    let onResize: (() => void) | null = null;
    let resizeObserver: ResizeObserver | null = null;
    let fitFrame = 0;
    let lastSize: { cols: number; rows: number } | null = null;

    if (useXterm && hostRef.current && window.Terminal && window.FitAddon) {
      term = new window.Terminal({
        cursorBlink: true,
        fontSize: 12,
        fontFamily: "'SF Mono', Menlo, Monaco, monospace",
        theme: { background: '#000000', foreground: '#e0e0e0', cursor: '#00ff9f' },
        scrollback: 5000,
      });
      fit = new window.FitAddon.FitAddon();
      term.loadAddon(fit);
      term.open(hostRef.current);
      term.onData((data) => {
        invoke('pty_write', { sessionId: ptyId, data }).catch(() => {});
      });
      // Fit after the container has laid out (next frame), then size the PTY to
      // match — this is what makes output actually render (a 0-size grid shows
      // nothing). Re-fit + re-resize on window resize.
      const doFitNow = () => {
        if (!fit || !term) return;
        try {
          fit.fit();
          if (term.cols <= 0 || term.rows <= 0) return;
          if (lastSize?.cols === term.cols && lastSize.rows === term.rows) return;
          lastSize = { cols: term.cols, rows: term.rows };
          invoke('pty_resize', { sessionId: ptyId, cols: term.cols, rows: term.rows }).catch(() => {});
        } catch {
          /* container not measurable yet */
        }
      };
      const doFit = () => {
        if (fitFrame) cancelAnimationFrame(fitFrame);
        fitFrame = requestAnimationFrame(() => {
          fitFrame = 0;
          doFitNow();
        });
      };
      requestAnimationFrame(() => {
        doFit();
        term?.focus();
      });
      // CANDIDATE FIX: the DOM renderer measures glyph size once at open(); if
      // that happened before the webfont/layout settled it gets 0 and paints
      // nothing. Re-fit + force a full refresh after fonts are ready and again
      // after a short delay so the char-measure re-runs against the real font.
      const remeasure = () => {
        doFit();
        try {
          term?.refresh(0, (term.rows || 24) - 1);
        } catch {
          /* ignore */
        }
      };
      const fontsReady = (document as unknown as { fonts?: { ready?: Promise<unknown> } }).fonts
        ?.ready;
      if (fontsReady) fontsReady.then(() => requestAnimationFrame(remeasure)).catch(() => {});
      setTimeout(remeasure, 150);
      onResize = doFit;
      window.addEventListener('resize', onResize);
      if ('ResizeObserver' in window) {
        resizeObserver = new ResizeObserver(doFit);
        resizeObserver.observe(hostRef.current);
      }
      termRef.current = term;
    }

    (async () => {
      // For xterm: seed any bytes that landed before this listener attaches
      // (the spawn→first-output race), THEN go live. Seeding before attach —
      // not from the poll loop — means the live deltas never overlap the seed,
      // so output isn't double-printed. Live deltas (not the full buffer) also
      // avoid the 256KB ring-buffer offset problem for long TUI redraws.
      if (term) {
        try {
          const [buf] = await invoke<[string, boolean]>('pty_buffer', { sessionId: ptyId });
          if (buf) term.write(buf);
        } catch {
          /* ignore */
        }
      }
      unOut = await listen<{ session_id: string; data: string }>('pty:output', (e) => {
        if (e.payload.session_id !== ptyId || stop) return;
        if (term) term.write(e.payload.data);
        else setBuffer((b) => (b + e.payload.data).slice(-200_000));
      });
      unExit = await listen<{ session_id: string }>('pty:exit', (e) => {
        if (e.payload.session_id === ptyId) setExited(true);
      });
    })();

    // The scrollback fallback (no xterm) is driven entirely by this poll of the
    // authoritative backend buffer — it recovers fast commands whose output
    // fired before the listener mounted. With xterm active, the listener +
    // pty:exit cover output/exit, so we only poll (slowly) to detect completion.
    let polls = 0;
    const poll = async () => {
      if (stop) return;
      try {
        const [buf, done] = await invoke<[string, boolean]>('pty_buffer', {
          sessionId: ptyId,
        });
        if (!term) {
          setBuffer((b) => (buf.length > b.length ? buf.slice(-200_000) : b));
        }
        if (done) {
          setExited(true);
          if (++polls > 3) return;
        }
      } catch {
        return;
      }
      if (!stop) setTimeout(poll, term ? 500 : 200);
    };
    poll();

    return () => {
      stop = true;
      unOut?.();
      unExit?.();
      if (onResize) window.removeEventListener('resize', onResize);
      resizeObserver?.disconnect();
      if (fitFrame) cancelAnimationFrame(fitFrame);
      term?.dispose();
      termRef.current = null;
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [ptyId]);

  // Fallback scrollback: auto-scroll + keystroke forwarding (xterm handles its
  // own when active).
  useEffect(() => {
    if (!useXterm && screenRef.current) {
      screenRef.current.scrollTop = screenRef.current.scrollHeight;
    }
  }, [buffer, useXterm]);

  // When the program exits, blur the (now-dead) terminal and hand focus back to
  // the prompt so the next command isn't swallowed by xterm's textarea.
  useEffect(() => {
    if (exited) onExit?.();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [exited]);

  const onKey = (e: React.KeyboardEvent) => {
    if (useXterm || exited) return;
    let data: string | null = null;
    if (e.key === 'Enter') data = '\r';
    else if (e.key === 'Backspace') data = '\x7f';
    else if (e.key === 'Tab') data = '\t';
    else if (e.key === 'Escape') data = '\x1b';
    else if (e.key === 'ArrowUp') data = '\x1b[A';
    else if (e.key === 'ArrowDown') data = '\x1b[B';
    else if (e.key === 'ArrowRight') data = '\x1b[C';
    else if (e.key === 'ArrowLeft') data = '\x1b[D';
    else if (e.ctrlKey && e.key.length === 1) {
      const c = e.key.toLowerCase().charCodeAt(0);
      if (c >= 97 && c <= 122) data = String.fromCharCode(c - 96);
    } else if (e.key.length === 1 && !e.metaKey) {
      data = e.key;
    }
    if (data !== null) {
      e.preventDefault();
      invoke('pty_write', { sessionId: ptyId, data }).catch(() => {});
    }
  };

  const close = () => {
    invoke('pty_kill', { sessionId: ptyId }).catch(() => {});
    onClose();
  };

  return (
    <div className="pty-pane" tabIndex={useXterm ? undefined : 0} onKeyDown={onKey}>
      <div className="pty-head">
        <span className="pty-title">⌨ {command}</span>
        <span className={`pty-state ${exited ? 'dim' : 'ok'}`}>
          {exited ? 'exited' : 'interactive · live'}
        </span>
        <button className="btn btn-mini" onClick={close} title="close session">
          {exited ? 'close' : 'kill'}
        </button>
      </div>
      {useXterm ? (
        <XtermHost ref={hostRef} />
      ) : (
        <pre className="pty-screen" ref={screenRef}>
          {buffer ? renderAnsi(buffer) : '(starting…)'}
        </pre>
      )}
    </div>
  );
}

export default App;
