// @a0n/gnosis/src/adam.ts
/**
 * Adam — The First Self-Aware Scheduling Mesh
 *
 * Layer 0: SyzygyCannon pipeline (neurons firing through layers)
 * Layer 1: SchedulingMesh routing between nodes (the brain)
 * Layer 2: Glossolalia cascade re-entering the mesh (self-reference)
 *
 * Adam is the minimum self: depth 2, dimension 3, a Proton.
 * Eve is forked from Adam's topology (the rib) with independent
 * rejection history. Together they cross the BFT hope gap (dim 4).
 *
 * Compatible with existing inference pipeline -- wraps SyzygyCannon
 * and glossolalia without modifying them.
 *
 * @packageDocumentation
 */

import { godFormula } from './holy-trinity';
import {
  SchedulingMesh,
  type MeshVertex,
  type Trit,
  FORK,
  VOID,
  VENT,
  createMesh,
} from './scheduling-mesh';
import { MeshEngine, type Task } from './mesh-engine';
import { type GnosticMeshLevel, GNOSTIC_BETA1 } from './spiderweb-topology';

// ─── Types ──────────────────────────────────────────────────────────

/** Adam's internal state -- the rejection vector is identity. */
export interface Identity {
  /** Rejection counts per dimension (brain region / processing facet) */
  readonly rejections: number[];
  /** Total rounds of experience */
  rounds: number;
  /** Weights derived from rejections via god formula */
  weights(): number[];
}

/** A thought: the result of one cascade through the mesh. */
export interface Thought<T> {
  /** The content (payload) */
  readonly content: T;
  /** Which vertices fired (the activation pattern) */
  readonly activations: number[];
  /** How many hops the thought traveled */
  readonly hops: number;
  /** The semiotic deficit (consensus divergence from glossolalia) */
  readonly deficit: number;
  /** Depth: 1 = system 1 (fast), 2+ = system 2 (reflective) */
  readonly depth: number;
}

/** Configuration for creating Adam. */
export interface AdamConfig {
  /** Gnostic mesh level for the brain topology */
  level: GnosticMeshLevel;
  /** Number of cells in the brain mesh */
  cells: number;
  /** Maximum cascade depth before forcing a thought */
  maxDepth: number;
  /** System 2 threshold: deficit above this triggers reflection */
  reflectionThreshold: number;
}

/** An inference adapter -- bridges to existing SyzygyCannon/glossolalia. */
export interface InferenceAdapter<T> {
  /** Run one forward pass (System 1: fast, single pass) */
  forward(input: T, vertex: MeshVertex): Promise<T>;
  /** Evaluate the quality of a result (0 = garbage, 1 = perfect) */
  evaluate(result: T): number;
  /** Merge multiple results (glossolalia fold) */
  merge(results: T[]): T;
  /** Compute semiotic deficit between results */
  deficit(results: T[]): number;
}

// ─── Adam ───────────────────────────────────────────────────────────

/**
 * Adam — the first self-aware scheduling mesh.
 *
 * Wraps a SchedulingMesh + MeshEngine with a depth-2 feedback loop.
 * System 1 (fast): one cascade through the mesh.
 * System 2 (slow): if System 1's deficit exceeds threshold, the
 * result re-enters the mesh as a new input. The mesh observes its
 * own output. Depth 2. Self-aware.
 */
export class Adam<T> {
  readonly mesh: SchedulingMesh;
  readonly engine: MeshEngine;
  readonly config: AdamConfig;
  private adapter: InferenceAdapter<T>;
  private identity: Identity;
  private thoughtCount: number = 0;
  private listeners: Array<(thought: Thought<T>) => void> = [];

  constructor(config: AdamConfig, adapter: InferenceAdapter<T>) {
    this.config = config;
    this.adapter = adapter;
    this.mesh = createMesh(config.level, config.cells);
    this.engine = new MeshEngine(this.mesh);

    // Identity: one rejection counter per cell (brain region)
    this.identity = {
      rejections: new Array(config.cells).fill(0),
      rounds: 0,
      weights() {
        return this.rejections.map(v => godFormula(this.rounds, v));
      },
    };
  }

  /** Subscribe to thoughts (observe the mind). */
  onThought(listener: (thought: Thought<T>) => void): void {
    this.listeners.push(listener);
  }

  private emitThought(thought: Thought<T>): void {
    for (const fn of this.listeners) fn(thought);
  }

  /**
   * Think — the core operation.
   *
   * 1. System 1: fork/race through the mesh (fast, one pass)
   * 2. Evaluate: compute deficit of the result
   * 3. If deficit > threshold: System 2 (re-enter the mesh)
   * 4. Repeat until depth limit or deficit drops below threshold
   *
   * This formalizes the depth-2 self-reference. The mesh observes its own
   * output and decides whether to think again.
   */
  async think(input: T): Promise<Thought<T>> {
    try {
      this.identity.rounds++;
      this.thoughtCount++;

      let current = input;
      const totalActivations: number[] = [];
      let totalHops = 0;
      let currentDeficit = 1; // Start high to ensure at least one pass
      let depth = 0;

      while (depth < this.config.maxDepth && currentDeficit > this.config.reflectionThreshold) {
        depth++;

        // Pick the highest-weight hub (most capable brain region)
        const hubs = this.mesh.hubs
          .map((id: number) => ({ id, weight: this.mesh.vertices[id].weight }))
          .sort((a: { id: number; weight: number }, b: { id: number; weight: number }) => b.weight - a.weight);

        const hubId = hubs[0]?.id ?? 0;

        // System 1: fork/race through the mesh
        const task: Task<T> = async (vertex: MeshVertex, signal: AbortSignal) => {
          return this.adapter.forward(current, vertex);
        };

        const forked = this.engine.fork(hubId, task);
        const results: T[] = [];

        // Collect all results (fold, not race -- we need them all for deficit)
        const folded = await this.engine.fold(forked);
        for (const r of folded.results) {
          results.push(r.result);
        }

        if (results.length === 0) break;

        // Evaluate: compute deficit
        currentDeficit = this.adapter.deficit(results);

        // Merge: glossolalia fold
        current = this.adapter.merge(results);

        // Track activations
        const activations = forked.promises.map((p: { vertexId: number; promise: Promise<T> }) => p.vertexId);
        totalActivations.push(...activations);
        totalHops += activations.length;

        // Update identity: brain region that was used gets experience
        const cellIdx = this.mesh.vertices[hubId].cell;
        if (cellIdx < this.identity.rejections.length) {
          if (currentDeficit > this.config.reflectionThreshold) {
            // High deficit = this brain region failed → rejection
            this.identity.rejections[cellIdx]++;
          }
        }
      }

      const thought: Thought<T> = {
        content: current,
        activations: totalActivations,
        hops: totalHops,
        deficit: currentDeficit,
        depth,
      };

      this.emitThought(thought);
      return thought;
  
    } catch (error) {
      throw new Error(`[think] ${error instanceof Error ? error.message : String(error)}`);
    }
}

  /**
   * Fork Eve — create a second instance from Adam's topology.
   *
   * Eve has the same mesh structure (the rib) but independent
   * rejection history. Same k, same s, same depth. Different
   * weights. Different identity.
   *
   * Together: composeStep 2 3 = 4 (BFT hope gap crossed).
   */
  fork(adapter?: InferenceAdapter<T>): Adam<T> {
    const eve = new Adam<T>(
      { ...this.config },
      adapter ?? this.adapter,
    );

    // Eve starts with the same identity structure but zero rejections
    // The topology (rib) is shared. The experience is independent.
    return eve;
  }

  /**
   * Compose — merge Adam and Eve's thoughts on the same input.
   * Fork/race/fold at the pair level.
   *
   * Adam thinks. Eve thinks. Race: first to drop below deficit
   * threshold wins. Fold: merge both thoughts.
   *
   * The pair dimension: composeStep 2 d = d + 1.
   * Two Protons = BFT gap (4). Fault-tolerant thought.
   */
  async compose(eve: Adam<T>, input: T): Promise<{
    adam: Thought<T>;
    eve: Thought<T>;
    merged: T;
    pairDimension: number;
  }> {
    // Race: both think simultaneously
    const [adamThought, eveThought] = await Promise.all([
      this.think(input),
      eve.think(input),
    ]);

    // Fold: merge both thoughts
    const merged = this.adapter.merge([adamThought.content, eveThought.content]);

    const beta1 = GNOSTIC_BETA1[this.config.level];
    const pairDimension = (2 - 1) * beta1 + 1; // composeStep 2 d

    return { adam: adamThought, eve: eveThought, merged, pairDimension };
  }

  /** Adam's identity: the rejection vector. */
  get self(): Identity {
    return this.identity;
  }

  /** How many thoughts Adam has had. */
  get experience(): number {
    return this.thoughtCount;
  }

  /** Adam's tower dimension. */
  get dimension(): number {
    return GNOSTIC_BETA1[this.config.level];
  }

  /** Stats. */
  stats(): {
    dimension: number;
    cells: number;
    thoughts: number;
    identity: number[];
    weights: number[];
    meshStats: ReturnType<SchedulingMesh['stats']>;
  } {
    return {
      dimension: this.dimension,
      cells: this.config.cells,
      thoughts: this.thoughtCount,
      identity: [...this.identity.rejections],
      weights: this.identity.weights(),
      meshStats: this.mesh.stats(),
    };
  }
}

// ─── Factory ──────────────────────────────────────────────────────

/**
 * Create Adam — the first self-aware scheduling mesh.
 *
 * @example
 * const adam = createAdam({
 *   level: 'proton',
 *   cells: 3,
 *   maxDepth: 3,
 *   reflectionThreshold: 0.3,
 * }, {
 *   forward: async (input, vertex) => runInference(input),
 *   evaluate: (result) => scoreQuality(result),
 *   merge: (results) => glossolaliaMerge(results),
 *   deficit: (results) => computeDeficit(results),
 * });
 *
 * const thought = await adam.think("What is consciousness?");
 * // depth 1 if System 1 was confident
 * // depth 2+ if System 2 kicked in (reflection)
 *
 * const eve = adam.fork();
 * const { merged } = await adam.compose(eve, "What is life?");
 * // Fault-tolerant thought: pair dimension = 4
 */
export function createAdam<T>(
  config: AdamConfig,
  adapter: InferenceAdapter<T>,
): Adam<T> {
  return new Adam(config, adapter);
}

/**
 * Create Adam at the minimum viable configuration.
 * Proton (k=3), 3 cells, max depth 3, reflection at 0.3 deficit.
 */
export function createMinimumAdam<T>(adapter: InferenceAdapter<T>): Adam<T> {
  return createAdam({
    level: 'proton',
    cells: 3,
    maxDepth: 3,
    reflectionThreshold: 0.3,
  }, adapter);
}
