// @a0n/gnosis/src/adt-handlers.ts
/**
 * ADT GG Topology Handlers
 *
 * Handlers for sum type definition, variant creation, exhaustive matching,
 * and type registry operations within GG topology execution.
 */

import type { GnosisRegistry } from './runtime/registry';
import {
  type SumTypeDefinition,
  type ProductTypeDefinition,
  type Variant,
  variant,
  defineSumType,
  defineProductType,
  createRecord,
  matchExhaustive,
  matchPartial,
  checkExhaustiveness,
  validateNodeExhaustiveness,
  TypeRegistry,
  BUILTIN_SUM_TYPES,
} from './adt';

/**
 * Handles the gnosis register Adt Handlers workflow.
 */
export function registerAdtHandlers(registry: GnosisRegistry): void {
  const typeRegistry = new TypeRegistry();

  // Define a new sum type
  registry.register('SumTypeDefiner', async (payload, props) => {
    const p = payload as Record<string, unknown> | undefined;
    const name = (props.name ?? p?.name ?? '') as string;
    const cases = ((props.cases ?? p?.cases ?? '') as string)
      .split(',')
      .map((s: string) => s.trim())
      .filter(Boolean);
    const sealed = props.sealed !== 'false';

    if (!name || cases.length === 0) {
      return { error: 'Name and cases required' };
    }

    const def = defineSumType(name, cases, sealed);
    typeRegistry.registerSumType(def);
    return { registered: true, name, cases, sealed };
  });

  // Create a variant value
  registry.register('VariantCreator', async (payload, props) => {
    const tag = props.tag ?? '';
    const value = payload;
    return variant(tag, value);
  });

  // Check exhaustiveness of cases against a type
  registry.register('ExhaustivenessChecker', async (payload, props) => {
    const typeName = props.type ?? '';
    const cases = (props.cases ?? '')
      .split(',')
      .map((s: string) => s.trim())
      .filter(Boolean);

    const typeDef = typeRegistry.getSumType(typeName);
    if (!typeDef) {
      return { error: `Unknown sum type '${typeName}'` };
    }

    return checkExhaustiveness(typeDef, cases);
  });

  // Validate node exhaustiveness (topology-level check)
  registry.register('NodeExhaustivenessValidator', async (payload, props) => {
    const nodeId = props.nodeId ?? '';
    const typeName = props.type ?? '';
    const edgeCases = (props.edgeCases ?? '')
      .split(',')
      .map((s: string) => s.trim())
      .filter(Boolean);

    const typeDef = typeRegistry.getSumType(typeName);
    if (!typeDef) {
      return { error: `Unknown sum type '${typeName}'` };
    }

    return validateNodeExhaustiveness(nodeId, typeDef, edgeCases);
  });

  // Exhaustive match on a variant
  registry.register('ExhaustiveMatch', async (payload, props) => {
    const typeName = props.type ?? '';
    const typeDef = typeRegistry.getSumType(typeName);
    if (!typeDef) {
      return { error: `Unknown sum type '${typeName}'` };
    }

    const v = payload as Variant;
    // Arms are provided as JSON in props.arms
    const armsRaw = props.arms ? JSON.parse(props.arms) : {};
    const arms: Record<string, (v: unknown) => unknown> = {};
    for (const [tag, result] of Object.entries(armsRaw)) {
      arms[tag] = () => result;
    }

    try {
      const result = matchExhaustive(typeDef, v, arms);
      return { matched: true, result, tag: v.tag };
    } catch (e: any) {
      return { matched: false, error: e.message };
    }
  });

  // List all registered types
  registry.register('TypeRegistryLister', async () => {
    return {
      sumTypes: typeRegistry.allSumTypes().map((t) => ({
        name: t.name,
        cases: t.cases,
        sealed: t.sealed,
      })),
      productTypes: typeRegistry.allProductTypes().map((t) => ({
        name: t.name,
        fields: t.fields,
      })),
    };
  });
}

export const ADT_BUILTIN_TYPES = Object.keys(BUILTIN_SUM_TYPES);
