// @a0n/gnosis/src/aeon-flux-site-witness.ts
import { parseGgProgram, type GgProgram } from '@a0n/aeon-logic';

import { lowerUfcsSource } from './ufcs';

type ProgramNode = GgProgram['nodes'][number];
type ProgramEdge = GgProgram['edges'][number];

export type AeonFluxSiteAdequacyRegime =
  | 'lossless_transport_exists'
  | 'tight_lossless_transport_exists'
  | 'collision_and_information_loss_forced';

export type AeonFluxSiteAdequacyTheoremRef =
  | 'aeon_flux_site_nonpositive_deficit_iff_lossless_transport'
  | 'aeon_flux_site_zero_deficit_iff_tight_lossless_transport'
  | 'aeon_flux_site_positive_deficit_forces_collision_and_information_loss';

export interface AeonFluxSiteWitness {
  readonly primitive: 'WallingtonRotation';
  readonly primitiveId: string;
  readonly ingressNodeId: string;
  readonly schedulerNodeId: string;
  readonly stageAlignedNodeId: string;
  readonly outputNodeId: string;
  readonly chunkNodeIds: readonly string[];
  readonly stageNodeIds: readonly string[];
  readonly pathCount: number;
  readonly streamCount: number;
  readonly deficit: number;
  readonly regime: AeonFluxSiteAdequacyRegime;
  readonly theoremRef: AeonFluxSiteAdequacyTheoremRef;
}

function requireNode(
  program: GgProgram,
  nodeId: string,
  label: string
): ProgramNode {
  try {
    const node = program.nodes.find((candidate) => candidate.id === nodeId);
    if (!node) {
      throw new Error(`Expected ${label} node "${nodeId}" in lowered GG program`);
    }

    return node;

  } catch (error) {
    throw new Error(`[requireNode] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function requireSingleEdge(
  program: GgProgram,
  predicate: (edge: ProgramEdge) => boolean,
  description: string
): ProgramEdge {
  try {
    const matches = program.edges.filter(predicate);
    if (matches.length !== 1) {
      throw new Error(
        `Expected exactly one ${description} edge, found ${matches.length}`
      );
    }

    const [edge] = matches;
    if (!edge) {
      throw new Error(`Expected ${description} edge to exist`);
    }

    return edge;

  } catch (error) {
    throw new Error(`[requireSingleEdge] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function parseRequiredPositiveInt(
  rawValue: string | undefined,
  description: string
): number {
  try {
    const parsed = Number(rawValue);
    if (!Number.isInteger(parsed) || parsed <= 0) {
      throw new Error(
        `Expected positive integer ${description}, got "${rawValue}"`
      );
    }

    return parsed;

  } catch (error) {
    throw new Error(`[parseRequiredPositiveInt] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function collectStageNodeIds(
  program: GgProgram,
  stageAlignedNodeId: string,
  outputNodeId: string
): readonly string[] {
  try {
    const stageNodeIds: string[] = [];
    const seen = new Set<string>([stageAlignedNodeId]);
    let currentNodeId = stageAlignedNodeId;

    while (true) {
      const nextEdges = program.edges.filter(
        (edge) =>
          edge.type.toUpperCase() === 'PROCESS' &&
          edge.sourceIds.length === 1 &&
          edge.sourceIds[0] === currentNodeId
      );

      if (nextEdges.length !== 1) {
        throw new Error(
          `Expected exactly one PROCESS edge leaving "${currentNodeId}" in Wallington stage chain`
        );
      }

      const [nextEdge] = nextEdges;
      const [nextNodeId] = nextEdge?.targetIds ?? [];
      if (!nextNodeId) {
        throw new Error(
          `Expected PROCESS edge leaving "${currentNodeId}" to target one node`
        );
      }

      if (nextNodeId === outputNodeId) {
        return stageNodeIds;
      }

      if (seen.has(nextNodeId)) {
        throw new Error(
          `Detected cycle while walking Wallington stage chain at "${nextNodeId}"`
        );
      }

      seen.add(nextNodeId);
      stageNodeIds.push(nextNodeId);
      currentNodeId = nextNodeId;
    }

  } catch (error) {
    throw new Error(`[collectStageNodeIds] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function classifyAdequacyRegime(
  pathCount: number,
  streamCount: number
): {
  readonly regime: AeonFluxSiteAdequacyRegime;
  readonly theoremRef: AeonFluxSiteAdequacyTheoremRef;
} {
  try {
    const deficit = pathCount - streamCount;
    if (deficit > 0) {
      return {
        regime: 'collision_and_information_loss_forced',
        theoremRef:
          'aeon_flux_site_positive_deficit_forces_collision_and_information_loss',
      };
    }

    if (deficit === 0 && pathCount > 0) {
      return {
        regime: 'tight_lossless_transport_exists',
        theoremRef: 'aeon_flux_site_zero_deficit_iff_tight_lossless_transport',
      };
    }

    return {
      regime: 'lossless_transport_exists',
      theoremRef: 'aeon_flux_site_nonpositive_deficit_iff_lossless_transport',
    };

  } catch (error) {
    throw new Error(`[classifyAdequacyRegime] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function synthesizeWallingtonRotationWitness(
  program: GgProgram,
  schedulerNode: ProgramNode
): AeonFluxSiteWitness {
  try {
    const schedulerSuffix = '__scheduler';
    if (!schedulerNode.id.endsWith(schedulerSuffix)) {
      throw new Error(
        `Expected Wallington scheduler id to end with "${schedulerSuffix}", found "${schedulerNode.id}"`
      );
    }

    const primitiveId = schedulerNode.id.slice(0, -schedulerSuffix.length);
    const ingressNodeId = `${primitiveId}__ingress`;
    const schedulerNodeId = schedulerNode.id;
    const stageAlignedNodeId = `${primitiveId}__stage_aligned`;
    const outputNode = requireNode(program, primitiveId, 'Wallington output');

    requireNode(program, ingressNodeId, 'Wallington ingress');
    requireNode(program, stageAlignedNodeId, 'Wallington stage-aligned');

    const forkEdge = requireSingleEdge(
      program,
      (edge) =>
        edge.type.toUpperCase() === 'FORK' &&
        edge.sourceIds.length === 1 &&
        edge.sourceIds[0] === schedulerNodeId,
      `FORK out of "${schedulerNodeId}"`
    );
    const foldEdge = requireSingleEdge(
      program,
      (edge) =>
        edge.type.toUpperCase() === 'FOLD' &&
        edge.targetIds.length === 1 &&
        edge.targetIds[0] === stageAlignedNodeId,
      `FOLD into "${stageAlignedNodeId}"`
    );

    const chunkNodeIds = forkEdge.targetIds;
    const pathCount = chunkNodeIds.length;
    const streamCount = parseRequiredPositiveInt(
      schedulerNode.properties.chunks ?? outputNode.properties.chunks,
      `Wallington chunk/stream count for "${primitiveId}"`
    );

    if (pathCount === 0) {
      throw new Error(
        `Expected Wallington site "${primitiveId}" to expose at least one chunk path`
      );
    }

    if (foldEdge.sourceIds.length !== pathCount) {
      throw new Error(
        `Expected FOLD into "${stageAlignedNodeId}" to preserve ${pathCount} chunk paths, found ${foldEdge.sourceIds.length}`
      );
    }

    const stageNodeIds = collectStageNodeIds(
      program,
      stageAlignedNodeId,
      outputNode.id
    );
    const { regime, theoremRef } = classifyAdequacyRegime(pathCount, streamCount);

    return {
      primitive: 'WallingtonRotation',
      primitiveId,
      ingressNodeId,
      schedulerNodeId,
      stageAlignedNodeId,
      outputNodeId: outputNode.id,
      chunkNodeIds,
      stageNodeIds,
      pathCount,
      streamCount,
      deficit: pathCount - streamCount,
      regime,
      theoremRef,
    };

  } catch (error) {
    throw new Error(`[synthesizeWallingtonRotationWitness] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Handles the gnosis synthesize Aeon Flux Site Witnesses workflow.
 */
export function synthesizeAeonFluxSiteWitnesses(
  program: GgProgram
): readonly AeonFluxSiteWitness[] {
  const wallingtonSchedulers = program.nodes.filter(
    (node) =>
      node.labels.includes('RotationScheduler') &&
      node.id.endsWith('__scheduler') &&
      program.nodes.some(
        (candidate) =>
          candidate.id === node.id.slice(0, -'__scheduler'.length) &&
          candidate.properties.primitive === 'WallingtonRotation'
      )
  );

  return wallingtonSchedulers.map((schedulerNode) =>
    synthesizeWallingtonRotationWitness(program, schedulerNode)
  );
}

/**
 * Handles the gnosis synthesize Aeon Flux Site Witnesses From Source workflow.
 */
export function synthesizeAeonFluxSiteWitnessesFromSource(
  source: string
): readonly AeonFluxSiteWitness[] {
  const program = parseGgProgram(lowerUfcsSource(source));
  return synthesizeAeonFluxSiteWitnesses(program);
}
