// @a0n/gnosis/src/analysis.ts
import {
  checkGgProgram,
  getGgTerminalNodeIds,
  parseGgProgram,
  type CheckerResult,
  type CheckerTopologyStats,
  type GgTopologyState,
} from '@a0n/aeon-logic';
import {
  Points,
  Series,
  graphEda,
  graphOutliers,
  type GraphEdaSummary,
  type GraphEdge,
  type GraphOutlierResult,
  type PointsDescription,
  type SeriesDescription,
} from '@a0n/twokeys';
import {
  inferCapabilitiesFromGgSource,
  summarizeCapabilityRequirements,
  validateCapabilitiesForTarget,
  type CapabilityIssue,
  type CapabilityRequirement,
  type HostCapability,
  type RuntimeTarget,
} from './capabilities/index';
import {
  certifyGnosisProgram,
  createUniverseAdmissionReport,
  type GnosisUniverseAdmissionReport,
  toLawCapabilitySummary,
} from './law';
import { lowerUfcsSource } from './ufcs';

const performance =
  globalThis.performance ??
  ({
    now: (): number => Date.now(),
  } satisfies Pick<Performance, 'now'>);

export interface GnosisLineMetrics {
  totalLines: number;
  nonEmptyLines: number;
  commentLines: number;
  topologyLines: number;
}

export interface GnosisTopologyMetrics {
  nodeCount: number;
  functionNodeCount: number;
  edgeCount: number;
  expandedEdgeCount: number;
  structuralBeta1: number;
  forkEdgeCount: number;
  raceEdgeCount: number;
  foldEdgeCount: number;
  ventEdgeCount: number;
  sliverEdgeCount: number;
  processEdgeCount: number;
  observeEdgeCount: number;
  maxBranchFactor: number;
  avgBranchFactor: number;
  cyclomaticApprox: number;
}

export interface GnosisQuantumMetrics {
  superpositionEdgeCount: number;
  collapseEdgeCount: number;
  collapseCoverage: number;
  collapseDeficit: number;
  sliverDensity: number;
  betaPressure: number;
  betaHeadroom: number;
  quantumIndex: number;
}

export const VALID_GNOSIS_STEERING_MODES = [
  'off',
  'report',
  'suggest',
  'apply',
] as const;

export type GnosisSteeringMode = (typeof VALID_GNOSIS_STEERING_MODES)[number];

export const DEFAULT_GNOSIS_STEERING_MODE: GnosisSteeringMode = 'suggest';

export type GnosisSteeringRegime = 'laminar' | 'transitional' | 'turbulent';
export type GnosisSteeringBias = 'lean-in' | 'neutral' | 'pull-back';

export type GnosisSteeringAction =
  | 'expand'
  | 'staggered-expand'
  | 'hold'
  | 'multiplex'
  | 'constrain';

export type GnosisCollapseCostAction =
  | 'none'
  | 'keep-multiplicity'
  | 'pay-vent'
  | 'pay-repair'
  | 'pay-vent-or-repair';

export interface GnosisSteeringTelemetry {
  /**
   * Observed wall-clock time in microCharleys.
   * A microCharley is one millisecond.
   */
  wallMicroCharleys: number | null;
  /** Observed CPU time in microCharleys. */
  cpuMicroCharleys: number | null;
  /** wall / cpu, when both are available. */
  wallToCpuRatio: number | null;
}

export interface GnosisFailureTrilemmaMetrics {
  nontrivialFork: boolean;
  deterministicCollapseCandidate: boolean;
  singleTerminalObserved: boolean;
  ventCostEdges: number;
  repairCostEdges: number;
  totalVentCost: number;
  totalRepairDebt: number;
  totalPaidCost: number;
  paidStageCount: number;
  collapseCostFloor: number;
  collapseCostMargin: number;
  prefixCostDeficit: number;
  zeroWasteCollapseRisk: boolean;
  contagiousRepairPressure: boolean;
  collapseCostAction: GnosisCollapseCostAction;
  freeCollapsePrefixRisk: boolean;
}

export interface GnosisSteeringStopwatch {
  wallStartedAtMs: number;
  cpuStartedAt: NodeJS.CpuUsage;
}

export interface GnosisSteeringEda {
  frontierWidths: SeriesDescription;
  layerOccupancy: SeriesDescription;
  wavefrontShape: PointsDescription;
  graph: GraphEdaSummary<string>;
  graphOutliers: GraphOutlierResult<string>[];
}

export interface GnosisSteeringMetrics {
  mode: GnosisSteeringMode;
  autoApplyEnabled: boolean;
  /** Steering actuators are not wired yet; `apply` is contract-only for now. */
  applySupported: boolean;
  /** Primary scalar statistic for frontier pessimism / underfill; canonical Wallace surface. */
  wallaceNumber: number;
  wallaceUnit: 'wally';
  /** Short-form alias for Wallace output. */
  wally: number;
  topologyDeficit: number;
  frontierFill: number;
  /** Descriptive underfill field kept alongside the canonical Wallace surface. */
  frontierDeficit: number;
  regime: GnosisSteeringRegime;
  /** Signed boundary-walk bias. Positive = lean into concurrency, negative = pull back. */
  boundaryWalkBias: number;
  /** Human-readable steering posture derived from the boundary walk. */
  leanInBias: GnosisSteeringBias;
  recommendedAction: GnosisSteeringAction | null;
  /** Deterministic collapse is never free: vent, pay repair, or keep multiplicity alive. */
  failureBoundary: GnosisFailureTrilemmaMetrics;
  eda: GnosisSteeringEda;
  telemetry: GnosisSteeringTelemetry;
}

export interface GnosisComplexityReport {
  fileCount: number;
  line: GnosisLineMetrics;
  topology: GnosisTopologyMetrics;
  quantum: GnosisQuantumMetrics;
  steering: GnosisSteeringMetrics;
  buleyNumber: number;
  correctness: CheckerResult<GgTopologyState>;
  capabilities: GnosisCapabilityReport;
  universe: GnosisUniverseAdmissionReport;
}

export interface GnosisCapabilityReport {
  target: RuntimeTarget;
  required: CapabilityRequirement[];
  requiredUnique: HostCapability[];
  issues: CapabilityIssue[];
  ok: boolean;
}

export interface GnosisAnalyzeOptions {
  target?: RuntimeTarget;
  steeringMode?: GnosisSteeringMode;
}

type WallaceAliasTarget = {
  readonly wallaceNumber: number;
};

function round2(value: number): number {
  return Math.round(value * 100) / 100;
}

function buildLineMetrics(source: string): GnosisLineMetrics {
  const lines = source.split('\n');
  const totalLines = lines.length;
  const nonEmptyLines = lines.filter((line) => line.trim().length > 0).length;
  const commentLines = lines.filter((line) =>
    line.trim().startsWith('//')
  ).length;
  const topologyLines = lines.filter((line) => line.includes('-[:')).length;

  return {
    totalLines,
    nonEmptyLines,
    commentLines,
    topologyLines,
  };
}

type ParsedGgProgram = ReturnType<typeof parseGgProgram>;

interface GnosisLayeredStageCost {
  ventCost: number;
  repairDebt: number;
  collapseEdges: number;
}

function expandGraphEdges(program: ParsedGgProgram): GraphEdge<string>[] {
  const expandedEdges: GraphEdge<string>[] = [];

  for (const edge of program.edges) {
    for (const sourceId of edge.sourceIds) {
      for (const targetId of edge.targetIds) {
        expandedEdges.push({
          from: sourceId,
          to: targetId,
          weight: 1,
        });
      }
    }
  }

  return expandedEdges;
}

function countConnectedComponents(program: ParsedGgProgram): number {
  try {
    const neighbors = new Map<string, Set<string>>();

    for (const node of program.nodes) {
      neighbors.set(node.id, new Set<string>());
    }

    for (const edge of program.edges) {
      for (const sourceId of edge.sourceIds) {
        const sourceNeighbors = neighbors.get(sourceId);
        if (!sourceNeighbors) {
          continue;
        }

        for (const targetId of edge.targetIds) {
          sourceNeighbors.add(targetId);
          const targetNeighbors = neighbors.get(targetId);
          if (targetNeighbors) {
            targetNeighbors.add(sourceId);
          }
        }
      }
    }

    let components = 0;
    const visited = new Set<string>();

    for (const nodeId of neighbors.keys()) {
      if (visited.has(nodeId)) {
        continue;
      }

      components += 1;
      const frontier = [nodeId];

      while (frontier.length > 0) {
        const currentNodeId = frontier.pop();
        if (!currentNodeId || visited.has(currentNodeId)) {
          continue;
        }

        visited.add(currentNodeId);
        const currentNeighbors = neighbors.get(currentNodeId);
        if (!currentNeighbors) {
          continue;
        }

        for (const neighborId of currentNeighbors) {
          if (!visited.has(neighborId)) {
            frontier.push(neighborId);
          }
        }
      }
    }

    return Math.max(1, components);

  } catch (error) {
    throw new Error(`[countConnectedComponents] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function buildNodeLayerIndex(program: ParsedGgProgram): Map<string, number> {
  try {
    const indegreeByNodeId = new Map<string, number>();
    const outgoingByNodeId = new Map<string, Set<string>>();

    for (const node of program.nodes) {
      indegreeByNodeId.set(node.id, 0);
      outgoingByNodeId.set(node.id, new Set<string>());
    }

    for (const edge of program.edges) {
      for (const sourceId of edge.sourceIds) {
        const outgoing = outgoingByNodeId.get(sourceId);
        if (!outgoing) {
          continue;
        }

        for (const targetId of edge.targetIds) {
          outgoing.add(targetId);
          indegreeByNodeId.set(
            targetId,
            (indegreeByNodeId.get(targetId) ?? 0) + 1
          );
        }
      }
    }

    const rootNodeIds = [...indegreeByNodeId.entries()]
      .filter(([, indegree]) => indegree === 0)
      .map(([nodeId]) => nodeId);
    const frontier =
      rootNodeIds.length > 0
        ? rootNodeIds.map((nodeId) => ({ nodeId, layer: 0 }))
        : program.nodes.map((node) => ({ nodeId: node.id, layer: 0 }));
    const layerByNodeId = new Map<string, number>();

    while (frontier.length > 0) {
      const current = frontier.shift();
      if (!current) {
        continue;
      }

      const knownLayer = layerByNodeId.get(current.nodeId);
      if (knownLayer !== undefined && knownLayer <= current.layer) {
        continue;
      }

      layerByNodeId.set(current.nodeId, current.layer);
      const outgoing = outgoingByNodeId.get(current.nodeId);
      if (!outgoing) {
        continue;
      }

      for (const targetId of outgoing) {
        frontier.push({ nodeId: targetId, layer: current.layer + 1 });
      }
    }

    return layerByNodeId;

  } catch (error) {
    throw new Error(`[buildNodeLayerIndex] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function buildTopologyMetrics(program: ParsedGgProgram): GnosisTopologyMetrics {
  const edgeTypes = program.edges.map((edge) => edge.type.toUpperCase());
  const forkEdgeCount = edgeTypes.filter((type) => type === 'FORK').length;
  const raceEdgeCount = edgeTypes.filter((type) => type === 'RACE').length;
  const foldEdgeCount = edgeTypes.filter(
    (type) => type === 'FOLD' || type === 'COLLAPSE'
  ).length;
  const ventEdgeCount = edgeTypes.filter(
    (type) => type === 'VENT' || type === 'TUNNEL'
  ).length;
  const sliverEdgeCount = edgeTypes.filter((type) => type === 'SLIVER').length;
  const processEdgeCount = edgeTypes.filter(
    (type) => type === 'PROCESS'
  ).length;
  const observeEdgeCount = edgeTypes.filter(
    (type) => type === 'OBSERVE'
  ).length;
  const forkWidths = program.edges
    .filter((edge) => edge.type.toUpperCase() === 'FORK')
    .map((edge) => edge.targetIds.length);
  const maxBranchFactor = Math.max(1, ...forkWidths);
  const avgBranchFactor =
    forkWidths.length === 0
      ? 1
      : round2(
          forkWidths.reduce((sum, width) => sum + width, 0) / forkWidths.length
        );
  const expandedEdgeCount = program.edges.reduce(
    (sum, edge) => sum + edge.sourceIds.length * edge.targetIds.length,
    0
  );
  const connectedComponents = countConnectedComponents(program);
  const structuralBeta1 = Math.max(
    0,
    expandedEdgeCount - program.nodes.length + connectedComponents
  );

  return {
    nodeCount: program.nodes.length,
    functionNodeCount: program.nodes.filter((node) => node.labels.length > 0)
      .length,
    edgeCount: program.edges.length,
    expandedEdgeCount,
    structuralBeta1,
    forkEdgeCount,
    raceEdgeCount,
    foldEdgeCount,
    ventEdgeCount,
    sliverEdgeCount,
    processEdgeCount,
    observeEdgeCount,
    maxBranchFactor,
    avgBranchFactor,
    // McCabe-style approximation for directed graph workflows.
    cyclomaticApprox: Math.max(
      1,
      program.edges.length - program.nodes.length + 2
    ),
  };
}

/**
 * Handles the gnosis classify Steering Regime workflow.
 */
export function classifySteeringRegime(
  wallaceNumber: number
): GnosisSteeringRegime {
  try {
    if (wallaceNumber >= 0.4) {
      return 'turbulent';
    }
    if (wallaceNumber >= 0.15) {
      return 'transitional';
    }
    return 'laminar';

  } catch (error) {
    throw new Error(`[classifySteeringRegime] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Handles the gnosis surface Steering Metrics workflow.
 */
export function surfaceSteeringMetrics(mode: GnosisSteeringMode): boolean {
  return mode !== 'off';
}

/**
 * Handles the gnosis surface Steering Recommendations workflow.
 */
export function surfaceSteeringRecommendations(
  mode: GnosisSteeringMode
): boolean {
  return mode === 'suggest' || mode === 'apply';
}

function resolveSteeringMode(
  steeringMode?: GnosisSteeringMode
): GnosisSteeringMode {
  return steeringMode ?? DEFAULT_GNOSIS_STEERING_MODE;
}

/**
 * Handles the gnosis recommend Steering Action workflow.
 */
export function recommendSteeringAction(
  topologyDeficit: number,
  wallaceNumber: number,
  regime: GnosisSteeringRegime
): GnosisSteeringAction {
  try {
    if (topologyDeficit > 0.5 && wallaceNumber >= 0.25) {
      return 'staggered-expand';
    }
    if (topologyDeficit > 0.5) {
      return 'expand';
    }
    if (topologyDeficit < -0.5 && wallaceNumber >= 0.25) {
      return 'constrain';
    }
    if (regime === 'turbulent') {
      return 'multiplex';
    }
    return 'hold';

  } catch (error) {
    throw new Error(`[recommendSteeringAction] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function clampBias(value: number): number {
  return round2(Math.max(-1, Math.min(1, value)));
}

/**
 * Handles the gnosis derive Boundary Walk Bias workflow.
 */
export function deriveBoundaryWalkBias(
  topologyDeficit: number,
  wallaceNumber: number,
  recommendedAction: GnosisSteeringAction
): number {
  try {
    let bias = 0;

    if (topologyDeficit > 0) {
      bias += Math.min(0.6, topologyDeficit * 0.5);
    } else if (topologyDeficit < 0) {
      bias -= Math.min(0.6, Math.abs(topologyDeficit) * 0.5);
    }

    switch (recommendedAction) {
      case 'expand':
        bias += 0.25;
        break;
      case 'staggered-expand':
        bias += 0.15;
        break;
      case 'multiplex':
        bias += Math.max(0.2, Math.min(0.45, wallaceNumber));
        break;
      case 'constrain':
        bias -= 0.25;
        break;
      default:
        break;
    }

    return clampBias(bias);

  } catch (error) {
    throw new Error(`[deriveBoundaryWalkBias] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Handles the gnosis classify Steering Bias workflow.
 */
export function classifySteeringBias(
  boundaryWalkBias: number
): GnosisSteeringBias {
  try {
    if (boundaryWalkBias >= 0.2) {
      return 'lean-in';
    }
    if (boundaryWalkBias <= -0.2) {
      return 'pull-back';
    }
    return 'neutral';

  } catch (error) {
    throw new Error(`[classifySteeringBias] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function buildSteeringEda(
  program: ParsedGgProgram,
  correctness: CheckerResult<GgTopologyState>
): GnosisSteeringEda {
  const frontierWidths =
    correctness.topology.frontierByLayer.length > 0
      ? [...correctness.topology.frontierByLayer]
      : [0];
  const peakFrontier = Math.max(1, ...frontierWidths);
  const occupancyByLayer = frontierWidths.map((width) => width / peakFrontier);
  const wavefrontPoints = frontierWidths.map((width, index) => [
    index + 1,
    width,
    occupancyByLayer[index] ?? 0,
  ]);
  const nodes = program.nodes.map((node) => node.id);
  const edges = expandGraphEdges(program);

  const outlierResults = graphOutliers(nodes, edges, { method: 'combined' }) as
    | GraphOutlierResult<string>
    | GraphOutlierResult<string>[];
  const graphOutlierList = (
    Array.isArray(outlierResults) ? outlierResults : [outlierResults]
  ) as GraphOutlierResult<string>[];

  return {
    frontierWidths: new Series({ data: frontierWidths }).describe(),
    layerOccupancy: new Series({ data: occupancyByLayer }).describe(),
    wavefrontShape: new Points({ data: wavefrontPoints }).describe(),
    graph: graphEda(nodes, edges, { directed: true }),
    graphOutliers: graphOutlierList,
  };
}

function buildFailureTrilemmaMetrics(
  program: ParsedGgProgram,
  topology: GnosisTopologyMetrics,
  correctness: CheckerResult<GgTopologyState>
): GnosisFailureTrilemmaMetrics {
  try {
    const frontierByLayer =
      correctness.topology.frontierByLayer.length > 0
        ? [...correctness.topology.frontierByLayer]
        : [1];
    const layerByNodeId = buildNodeLayerIndex(program);
    const stageCosts = Array.from(
      {
        length: Math.max(0, frontierByLayer.length - 1),
      },
      (): GnosisLayeredStageCost => ({
        ventCost: 0,
        repairDebt: 0,
        collapseEdges: 0,
      })
    );
    const maxStageIndex = Math.max(0, stageCosts.length - 1);

    for (const edge of program.edges) {
      const expandedWeight = edge.sourceIds.length * edge.targetIds.length;
      const sourceLayer = Math.min(
        maxStageIndex,
        Math.max(
          0,
          ...edge.sourceIds.map((sourceId) => layerByNodeId.get(sourceId) ?? 0)
        )
      );
      const stage = stageCosts[sourceLayer];
      if (!stage) {
        continue;
      }

      switch (edge.type.toUpperCase()) {
        case 'VENT':
        case 'TUNNEL':
          stage.ventCost += expandedWeight;
          break;
        case 'SLIVER':
          stage.repairDebt += expandedWeight;
          break;
        case 'FOLD':
        case 'COLLAPSE':
        case 'RACE':
          stage.collapseEdges += expandedWeight;
          break;
        default:
          break;
      }
    }

    const terminalNodeCount = getGgTerminalNodeIds(program).length;
    const nontrivialFork =
      topology.forkEdgeCount > 0 && topology.maxBranchFactor > 1;
    const deterministicCollapseCandidate =
      nontrivialFork &&
      (topology.foldEdgeCount > 0 || topology.raceEdgeCount > 0);
    const singleTerminalObserved = terminalNodeCount <= 1;
    const ventCostEdges = topology.ventEdgeCount;
    const repairCostEdges = topology.sliverEdgeCount;
    const totalVentCost = stageCosts.reduce(
      (sum, stage) => sum + stage.ventCost,
      0
    );
    const totalRepairDebt = stageCosts.reduce(
      (sum, stage) => sum + stage.repairDebt,
      0
    );
    const totalPaidCost = totalVentCost + totalRepairDebt;
    const paidStageCount = stageCosts.filter(
      (stage) => stage.ventCost > 0 || stage.repairDebt > 0
    ).length;
    const zeroWasteCollapseRisk =
      deterministicCollapseCandidate &&
      singleTerminalObserved &&
      ventCostEdges === 0 &&
      repairCostEdges === 0;
    const contagiousRepairPressure =
      nontrivialFork && ventCostEdges === 0 && repairCostEdges > 0;
    const maxFrontier = Math.max(1, ...frontierByLayer);
    const collapseCostFloor = deterministicCollapseCandidate
      ? Math.max(0, maxFrontier - 1)
      : 0;
    let cumulativePaidCost = 0;
    let peakFrontier = Math.max(1, frontierByLayer[0] ?? 1);
    let prefixCostDeficit = 0;

    for (let stageIndex = 0; stageIndex < stageCosts.length; stageIndex += 1) {
      const stage = stageCosts[stageIndex];
      if (stage === undefined) {
        continue;
      }
      const currentFrontier = frontierByLayer[stageIndex] ?? peakFrontier;
      const nextFrontier =
        frontierByLayer[stageIndex + 1] ?? frontierByLayer[stageIndex] ?? 1;
      peakFrontier = Math.max(peakFrontier, currentFrontier);
      cumulativePaidCost += stage.ventCost + stage.repairDebt;

      const collapsePressure =
        stage.collapseEdges > 0 || nextFrontier < peakFrontier;
      if (!collapsePressure) {
        continue;
      }

      const requiredPrefixCost = Math.max(0, peakFrontier - nextFrontier);
      prefixCostDeficit = Math.max(
        prefixCostDeficit,
        requiredPrefixCost - cumulativePaidCost
      );
    }

    const collapseCostMargin = totalPaidCost - collapseCostFloor;
    const freeCollapsePrefixRisk =
      deterministicCollapseCandidate && prefixCostDeficit > 0;

    let collapseCostAction: GnosisCollapseCostAction = 'none';
    if (deterministicCollapseCandidate) {
      if (ventCostEdges === 0 && repairCostEdges === 0) {
        collapseCostAction = 'keep-multiplicity';
      } else if (ventCostEdges > 0 && repairCostEdges > 0) {
        collapseCostAction = 'pay-vent-or-repair';
      } else if (ventCostEdges > 0) {
        collapseCostAction = 'pay-vent';
      } else {
        collapseCostAction = 'pay-repair';
      }
    }

    return {
      nontrivialFork,
      deterministicCollapseCandidate,
      singleTerminalObserved,
      ventCostEdges,
      repairCostEdges,
      totalVentCost,
      totalRepairDebt,
      totalPaidCost,
      paidStageCount,
      collapseCostFloor,
      collapseCostMargin,
      prefixCostDeficit,
      zeroWasteCollapseRisk,
      contagiousRepairPressure,
      collapseCostAction,
      freeCollapsePrefixRisk,
    };

  } catch (error) {
    throw new Error(`[buildFailureTrilemmaMetrics] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Creates the Empty Steering Telemetry.
 */
export function createEmptySteeringTelemetry(): GnosisSteeringTelemetry {
  return {
    wallMicroCharleys: null,
    cpuMicroCharleys: null,
    wallToCpuRatio: null,
  };
}

/**
 * Handles the gnosis start Steering Telemetry workflow.
 */
export function startSteeringTelemetry(): GnosisSteeringStopwatch {
  return {
    wallStartedAtMs: performance.now(),
    cpuStartedAt: process.cpuUsage(),
  };
}

/**
 * Handles the gnosis finish Steering Telemetry workflow.
 */
export function finishSteeringTelemetry(
  stopwatch: GnosisSteeringStopwatch
): GnosisSteeringTelemetry {
  const wallMicroCharleys = round2(
    performance.now() - stopwatch.wallStartedAtMs
  );
  const cpuUsage = process.cpuUsage(stopwatch.cpuStartedAt);
  const cpuMicroCharleys = round2((cpuUsage.user + cpuUsage.system) / 1000);

  return {
    wallMicroCharleys,
    cpuMicroCharleys,
    wallToCpuRatio:
      cpuMicroCharleys === 0
        ? null
        : round2(wallMicroCharleys / cpuMicroCharleys),
  };
}

function withLegacyWallaceAliases<T extends WallaceAliasTarget>(
  value: T
): T & Pick<GnosisSteeringMetrics, 'wally' | 'frontierDeficit'> {
  return Object.defineProperties(value, {
    wally: {
      get: (): number => value.wallaceNumber,
      enumerable: false,
      configurable: true,
    },
    frontierDeficit: {
      get: (): number => value.wallaceNumber,
      enumerable: false,
      configurable: true,
    },
  }) as T & Pick<GnosisSteeringMetrics, 'wally' | 'frontierDeficit'>;
}

function normalizeCheckerTopology(
  topology: CheckerTopologyStats
): CheckerTopologyStats {
  const {
    wally: _legacyWally,
    frontierDeficit: _legacyFrontierDeficit,
    ...canonicalTopology
  } = topology;
  return withLegacyWallaceAliases({
    ...canonicalTopology,
    wallaceNumber: topology.wallaceNumber,
  }) as CheckerTopologyStats;
}

function normalizeSteeringMetrics(
  steering: GnosisSteeringMetrics
): GnosisSteeringMetrics {
  const {
    wally: _legacyWally,
    frontierDeficit: _legacyFrontierDeficit,
    ...canonicalSteering
  } = steering;
  return withLegacyWallaceAliases({
    ...canonicalSteering,
    wallaceNumber: steering.wallaceNumber,
  }) as GnosisSteeringMetrics;
}

/**
 * Handles the gnosis with Steering Telemetry workflow.
 */
export function withSteeringTelemetry(
  steering: GnosisSteeringMetrics,
  telemetry: GnosisSteeringTelemetry
): GnosisSteeringMetrics {
  return normalizeSteeringMetrics({
    ...steering,
    telemetry,
  });
}

function buildSteeringMetrics(
  program: ParsedGgProgram,
  topology: GnosisTopologyMetrics,
  correctness: CheckerResult<GgTopologyState>,
  steeringMode?: GnosisSteeringMode
): GnosisSteeringMetrics {
  const mode = resolveSteeringMode(steeringMode);
  const frontierFill = round2(correctness.topology.frontierFill);
  const checkerTopology =
    correctness.topology as typeof correctness.topology & {
      wallaceNumber?: number;
      wally?: number;
    };
  const wallaceNumber = round2(
    checkerTopology.wallaceNumber ??
      checkerTopology.wally ??
      correctness.topology.frontierDeficit
  );
  const topologyDeficit = round2(
    topology.structuralBeta1 - correctness.topology.beta1
  );
  const regime = classifySteeringRegime(wallaceNumber);
  const recommendedAction = recommendSteeringAction(
    topologyDeficit,
    wallaceNumber,
    regime
  );
  const boundaryWalkBias = deriveBoundaryWalkBias(
    topologyDeficit,
    wallaceNumber,
    recommendedAction
  );
  const leanInBias = classifySteeringBias(boundaryWalkBias);
  const failureBoundary = buildFailureTrilemmaMetrics(
    program,
    topology,
    correctness
  );

  return normalizeSteeringMetrics({
    mode,
    autoApplyEnabled: mode === 'apply',
    applySupported: false,
    wallaceNumber,
    wallaceUnit: 'wally',
    wally: wallaceNumber,
    topologyDeficit,
    frontierFill,
    frontierDeficit: wallaceNumber,
    regime,
    boundaryWalkBias,
    leanInBias,
    recommendedAction: surfaceSteeringRecommendations(mode)
      ? recommendedAction
      : null,
    failureBoundary,
    eda: buildSteeringEda(program, correctness),
    telemetry: createEmptySteeringTelemetry(),
  });
}

function buildQuantumMetrics(
  topology: GnosisTopologyMetrics,
  correctness: CheckerResult<GgTopologyState>
): GnosisQuantumMetrics {
  const superpositionEdgeCount =
    topology.forkEdgeCount + topology.sliverEdgeCount;
  // OBSERVE is a collapse operation — reading forces superposition to resolve
  const collapseEdgeCount =
    topology.raceEdgeCount +
    topology.foldEdgeCount +
    topology.ventEdgeCount +
    topology.observeEdgeCount;
  const collapseCoverage =
    topology.forkEdgeCount === 0
      ? 1
      : round2(collapseEdgeCount / topology.forkEdgeCount);
  const collapseDeficit = Math.max(
    0,
    topology.forkEdgeCount - collapseEdgeCount
  );
  const sliverDensity =
    topology.edgeCount === 0
      ? 0
      : round2(topology.sliverEdgeCount / topology.edgeCount);
  const betaPressure = round2(
    topology.maxBranchFactor * Math.max(1, topology.forkEdgeCount) +
      topology.sliverEdgeCount * 1.5 +
      correctness.topology.beta1
  );
  const betaHeadroom = round2(Math.max(0, 10 - correctness.topology.beta1));
  const quantumIndex = round2(
    betaPressure +
      collapseDeficit * 2 +
      Math.max(0, 1 - collapseCoverage) * 4 +
      sliverDensity * 5
  );

  return {
    superpositionEdgeCount,
    collapseEdgeCount,
    collapseCoverage,
    collapseDeficit,
    sliverDensity,
    betaPressure,
    betaHeadroom,
    quantumIndex,
  };
}

function computeBuleyNumber(
  line: GnosisLineMetrics,
  topology: GnosisTopologyMetrics
): number {
  const sizeComponent = Math.log2(line.nonEmptyLines + 1);
  const branchComponent =
    topology.forkEdgeCount * 2.3 +
    topology.raceEdgeCount * 1.6 +
    topology.sliverEdgeCount * 1.4 +
    topology.maxBranchFactor * 1.1;
  const collapsePenalty =
    Math.max(
      0,
      topology.forkEdgeCount -
        (topology.foldEdgeCount +
          topology.raceEdgeCount +
          topology.ventEdgeCount)
    ) * 1.75;
  const shapeComponent = topology.cyclomaticApprox * 1.35;
  const functionDensity =
    topology.nodeCount > 0
      ? topology.functionNodeCount / topology.nodeCount
      : 0;

  return round2(
    sizeComponent +
      branchComponent +
      collapsePenalty +
      shapeComponent +
      functionDensity * 2
  );
}

/**
 * Formats the Gnosis Violations.
 */
export function formatGnosisViolations(
  result: CheckerResult<GgTopologyState>
): string[] {
  try {
    if (result.ok) {
      return [];
    }

    return result.violations.map((violation) => {
      const trace = violation.trace.map((step) => step.stateId).join(' -> ');
      if (trace.length === 0) {
        return `${violation.kind}:${violation.name} ${violation.message}`;
      }
      return `${violation.kind}:${violation.name} ${violation.message} trace=${trace}`;
    });

  } catch (error) {
    throw new Error(`[formatGnosisViolations] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Handles the gnosis analyze Gnosis Source workflow.
 */
export async function analyzeGnosisSource(
  source: string,
  options: GnosisAnalyzeOptions = {}
): Promise<GnosisComplexityReport> {
  const normalizedSource = lowerUfcsSource(source);
  const line = buildLineMetrics(source);
  const program = parseGgProgram(normalizedSource);
  const topology = buildTopologyMetrics(program);
  const target = options.target ?? 'agnostic';
  const rawCorrectness = await checkGgProgram(normalizedSource, {
    defaults: {
      maxDepth: 128,
      maxBeta1Exclusive: 512,
    },
  });
  const correctness: CheckerResult<GgTopologyState> = {
    ...rawCorrectness,
    topology: normalizeCheckerTopology(rawCorrectness.topology),
  };
  const quantum = buildQuantumMetrics(topology, correctness);
  const steering = buildSteeringMetrics(
    program,
    topology,
    correctness,
    options.steeringMode
  );
  const capabilityRequirements = inferCapabilitiesFromGgSource(source);
  const capabilitySummary = summarizeCapabilityRequirements(
    capabilityRequirements
  );
  const capabilityValidation = validateCapabilitiesForTarget(
    capabilityRequirements,
    target
  );
  const universeEvaluation = certifyGnosisProgram(program, {
    subject: 'gnosis-analysis',
    source: normalizedSource,
    capabilitySummary: toLawCapabilitySummary(
      capabilitySummary,
      capabilityValidation.target,
      capabilityValidation.issues.length
    ),
    metadata: {
      analyzer: 'analyzeGnosisSource',
      target: capabilityValidation.target,
    },
  });
  const universe = createUniverseAdmissionReport(universeEvaluation, {
    correctnessOk: correctness.ok,
    capabilityOk: capabilityValidation.ok,
    reasons: [
      ...(!correctness.ok ? formatGnosisViolations(correctness) : []),
      ...capabilityValidation.issues.map(
        (issue) =>
          `${issue.severity.toUpperCase()} capability ${issue.capability}: ${
            issue.message
          }`
      ),
    ],
  });

  return {
    fileCount: 1,
    line,
    topology,
    quantum,
    steering,
    buleyNumber: computeBuleyNumber(line, topology),
    correctness,
    capabilities: {
      target: capabilityValidation.target,
      required: capabilityValidation.required,
      requiredUnique: capabilityValidation.requiredUnique,
      issues: capabilityValidation.issues,
      ok: capabilityValidation.ok,
    },
    universe,
  };
}
