// @a0n/gnosis/src/arena-merge-gate.ts
/**
 * Single evaluation point for Arena 5 merge promotion: deterministic verification
 * bundle, optional cooling gate, optional rollback replay on metric windows.
 * Pure functions; no I/O.
 */

import { Verdict } from '@a0n/bitwise';

import { mergeAllowedWhenCooling, type CoolingState } from './cooling-policy';
import {
  RollbackController,
  type RollbackPolicy,
  type RollbackSample,
} from './socio-topology/rollback-controller';
import { verifyBundle } from './verify-bundle';

export interface ArenaMergeGateCoolingInput {
  readonly state: CoolingState;
}

export interface ArenaMergeGateRollbackReplayInput {
  readonly policy: RollbackPolicy;
  readonly samples: readonly RollbackSample[];
}

export interface ArenaMergeGateInput {
  readonly verificationBundle: unknown;
  /** When set, merge is blocked while `state.mode === 'cooling'`. */
  readonly cooling?: ArenaMergeGateCoolingInput;
  /** When set, replays samples on a fresh controller; any rollback blocks merge. */
  readonly rollbackReplay?: ArenaMergeGateRollbackReplayInput;
}

/**
 * The merge decision, expressed on the canonical {@link Verdict} alphabet so
 * Arena 5 shares the grid's one verdict type instead of a bare `ok: boolean`:
 *
 *   - Accept  (admit)      — clean merge: verification passed, no holds.
 *   - Abstain (quarantine) — SOFT HOLD: the merge is not wrong, the system is
 *                            just too hot right now (cooling). Re-check once
 *                            cooling clears; nothing structural failed.
 *   - Decline (reject)     — HARD FAIL: verification failed or a rollback
 *                            fired; the patch is structurally unmergeable.
 *
 * `ok` is retained for back-compat as the lossy sign-collapse: only `Accept`
 * is `ok: true`; both `Decline` and the neutral `Abstain` collapse to
 * `ok: false` (the bool cannot recover which one occurred — that is the whole
 * point of carrying `verdict`).
 */
export type ArenaMergeGateResult =
  | { ok: true; verdict: Verdict.Accept }
  | { ok: false; verdict: Verdict.Abstain | Verdict.Decline; reasons: readonly string[] };

/**
 * Handles the gnosis evaluate Arena Merge Gate workflow.
 */
export function evaluateArenaMergeGate(input: ArenaMergeGateInput): ArenaMergeGateResult {
  const v = verifyBundle(input.verificationBundle);
  if (v.ok === false) {
    return { ok: false, verdict: Verdict.Decline, reasons: [`verification_${v.reason}`] };
  }

  // Hard-fail reasons force Decline; cooling alone is a soft Abstain (quarantine).
  const hardReasons: string[] = [];
  const softReasons: string[] = [];

  if (input.cooling !== undefined && !mergeAllowedWhenCooling(input.cooling.state)) {
    softReasons.push('cooling_active');
  }
  if (input.rollbackReplay !== undefined) {
    const ctl = new RollbackController(input.rollbackReplay.policy);
    for (const s of input.rollbackReplay.samples) {
      const d = ctl.step(s);
      if (d.action === 'rollback') {
        hardReasons.push(`rollback_${d.reason}`);
        break;
      }
    }
  }

  if (hardReasons.length > 0) {
    return { ok: false, verdict: Verdict.Decline, reasons: [...hardReasons, ...softReasons] };
  }
  if (softReasons.length > 0) {
    return { ok: false, verdict: Verdict.Abstain, reasons: softReasons };
  }
  return { ok: true, verdict: Verdict.Accept };
}
