// @a0n/gnosis/src/betti-wasm-manifest.ts
/**
 * Betti → WASM artifact manifest (schema `betti-wasm/1`).
 *
 * The Betti/polyglot emitter writes a sidecar `<artifactId>.manifest.json`
 * (and embeds the same JSON in the wasm custom section
 * `gnosis-betti-manifest`) describing an emitted artifact. Execution fields
 * reuse {@link WasmExecutionManifest} names verbatim (`export_name`, `abi`,
 * `memory_export`, `alloc_export`, `result_length_export`, `free_export`) so
 * a validated manifest spreads directly into
 * {@link GnosisWasmExecutionSurface.register} via
 * {@link toRegisterWasmArtifactRequest}.
 *
 * Everything here fails closed: malformed manifests throw, they are never
 * coerced or partially accepted.
 */
import type { WasmExecutionAbi } from './wasm-execution-bridge';
import type { RegisterWasmArtifactRequest } from './wasm-execution';

export const BETTI_WASM_ARTIFACT_VERSION = 'betti-wasm/1' as const;

/** Wasm custom section carrying the embedded manifest JSON. */
export const BETTI_WASM_MANIFEST_CUSTOM_SECTION = 'gnosis-betti-manifest';

export type BettiWasmTier = 'certified-direct' | 'interpreter-embed';

export type BettiWasmImportKind = 'function' | 'table' | 'memory' | 'global';

export interface BettiWasmImport {
  readonly module: string;
  readonly name: string;
  readonly kind: BettiWasmImportKind;
}

export interface BettiWasmLimits {
  /** Hard cap on linear memory, in 64KiB wasm pages. */
  readonly memoryPagesMax: number;
  /** Optional fuel budget (loop-iteration guard) baked into the artifact. */
  readonly fuel?: number;
}

export interface BettiWasmManifest {
  readonly artifact_version: typeof BETTI_WASM_ARTIFACT_VERSION;
  /** 16-hex fnv1a64 topology fingerprint (tag `monster-aot-wasm-v1`). */
  readonly artifactId: string;
  readonly topologyHash: string;
  readonly sourceHash: string;
  readonly language: string;
  readonly filePath: string;
  readonly functionName: string;
  readonly certified_class: string;
  readonly tier: BettiWasmTier;

  // Execution fields — WasmExecutionManifest names, verbatim.
  readonly export_name: string;
  readonly abi: WasmExecutionAbi;
  readonly memory_export?: string;
  readonly alloc_export?: string;
  readonly result_length_export?: string;
  readonly free_export?: string;

  /** Tier-1 (`certified-direct`) artifacts must declare an empty import set. */
  readonly imports: readonly BettiWasmImport[];
  readonly pure: boolean;
  readonly limits: BettiWasmLimits;
  readonly byteLength: number;
  /** 64-hex sha256 of the artifact bytes. */
  readonly sha256: string;
  /** Decline path mirrors aot_c's exit(3) discipline: the module traps. */
  readonly decline_semantics: 'trap';
  /** Whether a `wasm-opt` post-pass ran over the emitted module. */
  readonly wasm_opt: boolean;
}

const WASM_ABIS: readonly WasmExecutionAbi[] = ['scalar-v1', 'json-v1'];
const WASM_TIERS: readonly BettiWasmTier[] = [
  'certified-direct',
  'interpreter-embed',
];
const WASM_IMPORT_KINDS: readonly BettiWasmImportKind[] = [
  'function',
  'table',
  'memory',
  'global',
];

const ARTIFACT_ID_PATTERN = /^[0-9a-f]{16}$/;
const SHA256_PATTERN = /^[0-9a-f]{64}$/;

function manifestError(message: string): Error {
  return new Error(`betti-wasm/1 manifest: ${message}`);
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function requireString(
  record: Record<string, unknown>,
  field: string
): string {
  const value = record[field];
  if (typeof value !== 'string' || value.length === 0) {
    throw manifestError(
      `required field '${field}' must be a non-empty string, got ${describe(value)}`
    );
  }
  return value;
}

function optionalString(
  record: Record<string, unknown>,
  field: string
): string | undefined {
  const value = record[field];
  if (value === undefined) {
    return undefined;
  }
  if (typeof value !== 'string' || value.length === 0) {
    throw manifestError(
      `optional field '${field}' must be a non-empty string when present, got ${describe(value)}`
    );
  }
  return value;
}

function requireBoolean(
  record: Record<string, unknown>,
  field: string
): boolean {
  const value = record[field];
  if (typeof value !== 'boolean') {
    throw manifestError(
      `required field '${field}' must be a boolean, got ${describe(value)}`
    );
  }
  return value;
}

function describe(value: unknown): string {
  if (value === undefined) {
    return 'undefined';
  }
  if (value === null) {
    return 'null';
  }
  if (typeof value === 'string') {
    return JSON.stringify(value.length > 64 ? `${value.slice(0, 61)}...` : value);
  }
  if (Array.isArray(value)) {
    return `array(${value.length})`;
  }
  return typeof value === 'object' ? 'object' : `${typeof value} ${String(value)}`;
}

function validateImports(value: unknown): readonly BettiWasmImport[] {
  if (!Array.isArray(value)) {
    throw manifestError(
      `required field 'imports' must be an array, got ${describe(value)}`
    );
  }

  return value.map((entry, index) => {
    if (!isRecord(entry)) {
      throw manifestError(
        `imports[${index}] must be an object {module, name, kind}, got ${describe(entry)}`
      );
    }
    const module = entry['module'];
    const name = entry['name'];
    const kind = entry['kind'];
    if (typeof module !== 'string' || module.length === 0) {
      throw manifestError(
        `imports[${index}].module must be a non-empty string, got ${describe(module)}`
      );
    }
    if (typeof name !== 'string' || name.length === 0) {
      throw manifestError(
        `imports[${index}].name must be a non-empty string, got ${describe(name)}`
      );
    }
    if (!WASM_IMPORT_KINDS.includes(kind as BettiWasmImportKind)) {
      throw manifestError(
        `imports[${index}].kind must be one of ${WASM_IMPORT_KINDS.join(' | ')}, got ${describe(kind)}`
      );
    }
    return { module, name, kind: kind as BettiWasmImportKind };
  });
}

function validateLimits(value: unknown): BettiWasmLimits {
  if (!isRecord(value)) {
    throw manifestError(
      `required field 'limits' must be an object {memoryPagesMax, fuel?}, got ${describe(value)}`
    );
  }

  const memoryPagesMax = value['memoryPagesMax'];
  if (!Number.isSafeInteger(memoryPagesMax) || (memoryPagesMax as number) <= 0) {
    throw manifestError(
      `limits.memoryPagesMax must be a positive integer, got ${describe(memoryPagesMax)}`
    );
  }

  const fuel = value['fuel'];
  if (fuel !== undefined && (!Number.isSafeInteger(fuel) || (fuel as number) <= 0)) {
    throw manifestError(
      `limits.fuel must be a positive integer when present, got ${describe(fuel)}`
    );
  }

  return {
    memoryPagesMax: memoryPagesMax as number,
    ...(fuel !== undefined ? { fuel: fuel as number } : {}),
  };
}

/**
 * Validate an untrusted value as a `betti-wasm/1` manifest. Throws a
 * descriptive error on any malformed input — fail closed, never coerce.
 */
export function validateBettiWasmManifest(value: unknown): BettiWasmManifest {
  if (!isRecord(value)) {
    throw manifestError(`expected an object, got ${describe(value)}`);
  }

  const artifactVersion = value['artifact_version'];
  if (artifactVersion !== BETTI_WASM_ARTIFACT_VERSION) {
    throw manifestError(
      `artifact_version must be '${BETTI_WASM_ARTIFACT_VERSION}', got ${describe(artifactVersion)}`
    );
  }

  const artifactId = requireString(value, 'artifactId');
  if (!ARTIFACT_ID_PATTERN.test(artifactId)) {
    throw manifestError(
      `artifactId must be 16 lowercase hex characters (fnv1a64 fingerprint), got ${describe(artifactId)}`
    );
  }

  const abi = value['abi'];
  if (!WASM_ABIS.includes(abi as WasmExecutionAbi)) {
    throw manifestError(
      `abi must be one of ${WASM_ABIS.map((tag) => `'${tag}'`).join(' | ')}, got ${describe(abi)}`
    );
  }

  const tier = value['tier'];
  if (!WASM_TIERS.includes(tier as BettiWasmTier)) {
    throw manifestError(
      `tier must be one of ${WASM_TIERS.map((tag) => `'${tag}'`).join(' | ')}, got ${describe(tier)}`
    );
  }

  const imports = validateImports(value['imports']);
  if (tier === 'certified-direct' && imports.length > 0) {
    throw manifestError(
      `tier 'certified-direct' requires an empty import section, got ${imports.length} import(s) ` +
        `(first: ${imports[0].module}.${imports[0].name})`
    );
  }

  const sha256 = requireString(value, 'sha256');
  if (!SHA256_PATTERN.test(sha256)) {
    throw manifestError(
      `sha256 must be 64 lowercase hex characters, got ${describe(sha256)}`
    );
  }

  const byteLength = value['byteLength'];
  if (!Number.isSafeInteger(byteLength) || (byteLength as number) <= 0) {
    throw manifestError(
      `byteLength must be a positive integer, got ${describe(byteLength)}`
    );
  }

  const declineSemantics = value['decline_semantics'];
  if (declineSemantics !== 'trap') {
    throw manifestError(
      `decline_semantics must be 'trap', got ${describe(declineSemantics)}`
    );
  }

  const memoryExport = optionalString(value, 'memory_export');
  const allocExport = optionalString(value, 'alloc_export');
  const resultLengthExport = optionalString(value, 'result_length_export');
  const freeExport = optionalString(value, 'free_export');

  return {
    artifact_version: BETTI_WASM_ARTIFACT_VERSION,
    artifactId,
    topologyHash: requireString(value, 'topologyHash'),
    sourceHash: requireString(value, 'sourceHash'),
    language: requireString(value, 'language'),
    filePath: requireString(value, 'filePath'),
    functionName: requireString(value, 'functionName'),
    certified_class: requireString(value, 'certified_class'),
    tier: tier as BettiWasmTier,
    export_name: requireString(value, 'export_name'),
    abi: abi as WasmExecutionAbi,
    ...(memoryExport !== undefined ? { memory_export: memoryExport } : {}),
    ...(allocExport !== undefined ? { alloc_export: allocExport } : {}),
    ...(resultLengthExport !== undefined
      ? { result_length_export: resultLengthExport }
      : {}),
    ...(freeExport !== undefined ? { free_export: freeExport } : {}),
    imports,
    pure: requireBoolean(value, 'pure'),
    limits: validateLimits(value['limits']),
    byteLength: byteLength as number,
    sha256,
    decline_semantics: 'trap',
    wasm_opt: requireBoolean(value, 'wasm_opt'),
  };
}

export interface BettiWasmArtifactBytes {
  /** Inline base64 module bytes. Exactly one of the two must be set. */
  readonly wasmBase64?: string;
  /** Path to the `.wasm` artifact on disk. Exactly one of the two must be set. */
  readonly wasmPath?: string;
}

/**
 * Spread a validated manifest into a {@link RegisterWasmArtifactRequest} for
 * {@link GnosisWasmExecutionSurface.register}. The manifest's `artifactId` is
 * always passed through explicitly so the registry never re-derives identity.
 */
export function toRegisterWasmArtifactRequest(
  manifest: BettiWasmManifest,
  bytes: BettiWasmArtifactBytes
): RegisterWasmArtifactRequest {
  const hasBase64 =
    typeof bytes.wasmBase64 === 'string' && bytes.wasmBase64.length > 0;
  const hasPath = typeof bytes.wasmPath === 'string' && bytes.wasmPath.length > 0;
  if (!hasBase64 && !hasPath) {
    throw manifestError(
      'registration requires artifact bytes: pass wasmBase64 or wasmPath'
    );
  }
  if (hasBase64 && hasPath) {
    throw manifestError(
      'registration bytes are ambiguous: pass wasmBase64 or wasmPath, not both'
    );
  }

  return {
    artifactId: manifest.artifactId,
    ...(hasBase64 ? { wasmBase64: bytes.wasmBase64 } : {}),
    ...(hasPath ? { wasmPath: bytes.wasmPath } : {}),
    exportName: manifest.export_name,
    abi: manifest.abi,
    ...(manifest.memory_export !== undefined
      ? { memoryExport: manifest.memory_export }
      : {}),
    ...(manifest.alloc_export !== undefined
      ? { allocExport: manifest.alloc_export }
      : {}),
    ...(manifest.result_length_export !== undefined
      ? { resultLengthExport: manifest.result_length_export }
      : {}),
    ...(manifest.free_export !== undefined
      ? { freeExport: manifest.free_export }
      : {}),
  };
}

function sectionError(message: string): Error {
  return new Error(`betti-wasm custom section: ${message}`);
}

function readLeb128U32(
  bytes: Uint8Array,
  offset: number,
  context: string
): { readonly value: number; readonly nextOffset: number } {
  let value = 0;
  let cursor = offset;
  for (let index = 0; index < 5; index += 1) {
    if (cursor >= bytes.length) {
      throw sectionError(
        `truncated LEB128 ${context} at byte offset ${cursor}`
      );
    }
    const byte = bytes[cursor];
    cursor += 1;
    value += (byte & 0x7f) * 2 ** (index * 7);
    if ((byte & 0x80) === 0) {
      if (value > 0xffffffff) {
        throw sectionError(`LEB128 ${context} exceeds u32 range`);
      }
      return { value, nextOffset: cursor };
    }
  }
  throw sectionError(`LEB128 ${context} exceeds 5 bytes (not a u32)`);
}

const WASM_HEADER = Uint8Array.from([
  0x00, 0x61, 0x73, 0x6d, // magic "\0asm"
  0x01, 0x00, 0x00, 0x00, // version 1
]);

/**
 * Scan a wasm binary's custom sections for `gnosis-betti-manifest` and return
 * its validated `betti-wasm/1` manifest. Returns `null` when the module has
 * no such section; throws on malformed binaries or invalid manifests (fail
 * closed). Pure section walking — no WebAssembly runtime, no external deps.
 *
 * NOTE: this walker tolerates the manifest section at ANY position, but the
 * emitter always appends `gnosis-betti-manifest` LAST (the trailing section).
 * That trailing-by-construction placement is what makes the core-exclusive
 * integrity split well-defined: `manifest.byteLength` marks where the core
 * module ends, so `wasmBytes[..byteLength]` is the core and everything after is
 * this section (see `verifyBettiWasmIntegrity` in `betti-wasm-store.ts`).
 */
export function parseBettiManifestCustomSection(
  wasmBytes: Uint8Array
): BettiWasmManifest | null {
  if (wasmBytes.length < WASM_HEADER.length) {
    throw sectionError(
      `module is ${wasmBytes.length} byte(s), shorter than the 8-byte wasm header`
    );
  }
  for (let index = 0; index < WASM_HEADER.length; index += 1) {
    if (wasmBytes[index] !== WASM_HEADER[index]) {
      throw sectionError(
        index < 4
          ? 'bad magic (expected \\0asm)'
          : 'unsupported wasm version (expected 1)'
      );
    }
  }

  const utf8 = new TextDecoder('utf-8', { fatal: true });
  let offset = WASM_HEADER.length;
  while (offset < wasmBytes.length) {
    const sectionId = wasmBytes[offset];
    offset += 1;

    const size = readLeb128U32(wasmBytes, offset, 'section size');
    const sectionStart = size.nextOffset;
    const sectionEnd = sectionStart + size.value;
    if (sectionEnd > wasmBytes.length) {
      throw sectionError(
        `section id ${sectionId} declares ${size.value} byte(s) but only ` +
          `${wasmBytes.length - sectionStart} remain`
      );
    }

    if (sectionId === 0) {
      const nameLength = readLeb128U32(
        wasmBytes,
        sectionStart,
        'custom section name length'
      );
      const nameEnd = nameLength.nextOffset + nameLength.value;
      if (nameEnd > sectionEnd) {
        throw sectionError(
          'custom section name overruns its section payload'
        );
      }
      let name: string;
      try {
        name = utf8.decode(wasmBytes.subarray(nameLength.nextOffset, nameEnd));
      } catch {
        throw sectionError('custom section name is not valid UTF-8');
      }

      if (name === BETTI_WASM_MANIFEST_CUSTOM_SECTION) {
        let manifestJson: string;
        try {
          manifestJson = utf8.decode(wasmBytes.subarray(nameEnd, sectionEnd));
        } catch {
          throw sectionError(
            `'${BETTI_WASM_MANIFEST_CUSTOM_SECTION}' payload is not valid UTF-8`
          );
        }
        let parsed: unknown;
        try {
          parsed = JSON.parse(manifestJson);
        } catch (error) {
          throw sectionError(
            `'${BETTI_WASM_MANIFEST_CUSTOM_SECTION}' payload is not valid JSON: ` +
              (error instanceof Error ? error.message : String(error))
          );
        }
        return validateBettiWasmManifest(parsed);
      }
    }

    offset = sectionEnd;
  }

  return null;
}
