// @a0n/gnosis/src/bitwise-cipher.ts
/**
 * Helix Cipher — TypeScript port from @a0n/bitwise (Rust)
 *
 * A multi-dimensional structural cipher that exploits FlowFrame geometry.
 * Output is valid FlowFrame data that decodes to plausible but wrong
 * multi-channel data without the key. Steganographic by construction.
 *
 * Three cipher variants:
 *   1. HelixCipherKey — static 54-dimension permutation (~2^237 keyspace)
 *   2. RotatingHelixKey — dual-permutation with helical rotation (~2^474 keyspace)
 *   3. syzygy_smush/unsmush — ChaCha20-Poly1305 AEAD via @a0n/bitwise WASM
 *
 * The helix ciphers are structural: they permute dimensions, rotate phases,
 * and XOR payloads. They are NOT a replacement for AEAD — they provide
 * steganographic confusion. Layer with smush for full confidentiality.
 *
 * @see open-source/bitwise/src/lib.rs (canonical Rust implementation)
 */

// ═══════════════════════════════════════════════════════════════════════════════
// Xorshift64 PRNG (deterministic, NOT cryptographically secure)
// ═══════════════════════════════════════════════════════════════════════════════

export class Xorshift64 {
  private state: bigint;

  constructor(seed: bigint) {
    this.state = seed === 0n ? 0x5DEECE66Dn : seed;
  }

  next(): bigint {
    let x = this.state;
    x ^= (x << 13n) & 0xFFFFFFFFFFFFFFFFn;
    x ^= x >> 7n;
    x ^= (x << 17n) & 0xFFFFFFFFFFFFFFFFn;
    this.state = x;
    return x;
  }

  nextByte(): number {
    return Number(this.next() & 0xFFn);
  }

  fillBytes(buf: Uint8Array): void {
    for (let i = 0; i < buf.length; i += 8) {
      const val = this.next();
      const end = Math.min(i + 8, buf.length);
      for (let j = i; j < end; j++) {
        buf[j] = Number((val >> BigInt((j - i) * 8)) & 0xFFn);
      }
    }
  }
}

// ═══════════════════════════════════════════════════════════════════════════════
// FlowFrame helpers
// ═══════════════════════════════════════════════════════════════════════════════

/** Encode a helix stream ID from dimension + chromosome/source */
export function encodeHelixStreamId(dimension: number, source: number): number {
  return ((dimension & 0xFF) << 8) | (source & 0xFF);
}

/** Decode a helix stream ID into [dimension, source] */
export function decodeHelixStreamId(streamId: number): [number, number] {
  return [(streamId >> 8) & 0xFF, streamId & 0xFF];
}

/** Read payload length from FlowFrame header bytes 7-9 */
function readPayloadLength(frame: Uint8Array, offset: number): number {
  const byte7 = readIndexedNumber(frame, offset + 7, 'readPayloadLength byte 7');
  const byte8 = readIndexedNumber(frame, offset + 8, 'readPayloadLength byte 8');
  const byte9 = readIndexedNumber(frame, offset + 9, 'readPayloadLength byte 9');
  return (byte7 << 16) | (byte8 << 8) | byte9;
}

/** Read u32 from frame at offset (big-endian) */
function readU32BE(frame: Uint8Array, offset: number): number {
  const byte0 = readIndexedNumber(frame, offset, 'readU32BE byte 0');
  const byte1 = readIndexedNumber(frame, offset + 1, 'readU32BE byte 1');
  const byte2 = readIndexedNumber(frame, offset + 2, 'readU32BE byte 2');
  const byte3 = readIndexedNumber(frame, offset + 3, 'readU32BE byte 3');
  return ((byte0 << 24) | (byte1 << 16) | (byte2 << 8) | byte3) >>> 0;
}

/** Write u32 to frame at offset (big-endian) */
function writeU32BE(frame: Uint8Array, offset: number, value: number): void {
  frame[offset] = (value >>> 24) & 0xFF;
  frame[offset + 1] = (value >>> 16) & 0xFF;
  frame[offset + 2] = (value >>> 8) & 0xFF;
  frame[offset + 3] = value & 0xFF;
}

function readIndexedNumber(values: ArrayLike<number>, index: number, context: string): number {
  const value = values[index];
  if (value === undefined) {
    throw new Error(context);
  }
  return value;
}

// ═══════════════════════════════════════════════════════════════════════════════
// Helix Cipher Key (static permutation)
// ═══════════════════════════════════════════════════════════════════════════════

/**
 * The helix cipher key. Contains all secret parameters needed
 * to encrypt/decrypt a 54-dimension FlowFrame stream.
 *
 * Keyspace: 54! permutations (~2^237) x 2^1728 phase x 2^1728 dilation
 *           x 2^64 XOR x 2^16 stream mask
 */
export interface HelixCipherKey {
  /** dimension_permutation[plaintext_dim] = ciphertext_dim */
  dimensionPermutation: Uint8Array; // 54 entries
  /** inverse_permutation[ciphertext_dim] = plaintext_dim */
  inversePermutation: Uint8Array; // 54 entries
  /** Per-dimension position offset */
  phaseOffsets: Uint32Array; // 54 entries
  /** Per-dimension dilation stride override */
  dilationStrides: Uint32Array; // 54 entries
  /** Seed for XOR payload confusion */
  xorSeed: bigint;
  /** Stream ID XOR mask */
  streamIdMask: number;
}

/**
 * Generate a HelixCipherKey from a 64-byte seed.
 *
 * The seed is expanded via PRNG into all key components.
 * In practice, this seed comes from a KDF (HKDF, Argon2, etc.)
 * applied to a passphrase or shared secret.
 */
export function helixKeyFromSeed(seed: Uint8Array): HelixCipherKey {
  try {
    if (seed.length < 64) {
      throw new Error(`Helix key seed must be at least 64 bytes, got ${seed.length}`);
    }

    // Derive master PRNG state from seed
    let masterState = 0n;
    for (let i = 0; i < 64; i += 8) {
      let val = 0n;
      for (let j = 0; j < 8 && i + j < 64; j++) {
        val |= BigInt(readIndexedNumber(seed, i + j, 'helixKeyFromSeed seed byte')) << BigInt(j * 8);
      }
      masterState ^= val;
    }
    const rng = new Xorshift64(masterState);

    // Fisher-Yates shuffle for dimension permutation
    const perm = new Uint8Array(54);
    for (let i = 0; i < 54; i++) perm[i] = i;
    for (let i = 53; i >= 1; i--) {
      const j = Number(rng.next() % BigInt(i + 1));
      const tmp = readIndexedNumber(perm, i, 'helixKeyFromSeed perm[i]');
      const swap = readIndexedNumber(perm, j, 'helixKeyFromSeed perm[j]');
      perm[i] = swap;
      perm[j] = tmp;
    }

    // Compute inverse permutation
    const inv = new Uint8Array(54);
    for (let i = 0; i < 54; i++) {
      inv[readIndexedNumber(perm, i, 'helixKeyFromSeed inverse permutation')] = i;
    }

    // Phase offsets
    const phaseOffsets = new Uint32Array(54);
    for (let i = 0; i < 54; i++) {
      phaseOffsets[i] = Number(rng.next() & 0xFFFFFFFFn);
    }

    // Dilation strides (1-256 range)
    const dilationStrides = new Uint32Array(54);
    for (let i = 0; i < 54; i++) {
      dilationStrides[i] = Number((rng.next() % 256n) + 1n);
    }

    const xorSeed = rng.next();
    const streamIdMask = Number(rng.next() & 0xFFFFn);

    return {
      dimensionPermutation: perm,
      inversePermutation: inv,
      phaseOffsets,
      dilationStrides,
      xorSeed,
      streamIdMask,
    };

  } catch (error) {
    throw new Error(`[helixKeyFromSeed] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Encrypt a single FlowFrame using the helix cipher.
 *
 * Applies: dimension permutation, phase rotation, payload XOR, stream ID scramble.
 * Output is a valid FlowFrame with altered stream_id, position, and payload.
 */
export function helixEncryptFrame(frame: Uint8Array, key: HelixCipherKey): Uint8Array {
  try {
    if (frame.length < 10) return new Uint8Array(frame);

    const out = new Uint8Array(frame);

    // Decode stream_id
    const streamId = (readIndexedNumber(frame, 0, 'helixEncryptFrame streamId byte 0') << 8) |
      readIndexedNumber(frame, 1, 'helixEncryptFrame streamId byte 1');
    const dimension = (streamId >> 8) & 0xFF;
    const source = streamId & 0xFF;

    // 1. Dimension permutation
    const cipherDim =
      dimension < 54
        ? readIndexedNumber(key.dimensionPermutation, dimension, 'helixEncryptFrame dimension permutation')
        : dimension;

    // 2. Phase rotation
    const position = readU32BE(frame, 2);
    const offset =
      dimension < 54
        ? readIndexedNumber(key.phaseOffsets, dimension, 'helixEncryptFrame phase offset')
        : 0;
    const rotatedPos = (position + offset) >>> 0;
    writeU32BE(out, 2, rotatedPos);

    // 3. Payload XOR
    const payloadLen = readPayloadLength(frame, 0);
    const xorRng = new Xorshift64(key.xorSeed ^ (BigInt(cipherDim) * 0x9E3779B97F4A7C15n));
    const end = Math.min(10 + payloadLen, out.length);
    for (let i = 10; i < end; i++) {
      const current = readIndexedNumber(out, i, 'helixEncryptFrame payload byte');
      out[i] = current ^ xorRng.nextByte();
    }

    // 4. Stream ID scramble
    const newStreamId = encodeHelixStreamId(cipherDim, source) ^ key.streamIdMask;
    out[0] = (newStreamId >> 8) & 0xFF;
    out[1] = newStreamId & 0xFF;

    return out;

  } catch (error) {
    throw new Error(`[helixEncryptFrame] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Decrypt a single FlowFrame encrypted with helixEncryptFrame.
 * Reverses all cipher layers in reverse order.
 */
export function helixDecryptFrame(frame: Uint8Array, key: HelixCipherKey): Uint8Array {
  try {
    if (frame.length < 10) return new Uint8Array(frame);

    const out = new Uint8Array(frame);

    // 4. Reverse stream ID scramble
    const scrambledId = (readIndexedNumber(frame, 0, 'helixDecryptFrame streamId byte 0') << 8) |
      readIndexedNumber(frame, 1, 'helixDecryptFrame streamId byte 1');
    const unscrambledId = scrambledId ^ key.streamIdMask;
    const cipherDim = (unscrambledId >> 8) & 0xFF;
    const source = unscrambledId & 0xFF;

    // 1. Reverse dimension permutation
    const plainDim =
      cipherDim < 54
        ? readIndexedNumber(key.inversePermutation, cipherDim, 'helixDecryptFrame inverse permutation')
        : cipherDim;

    // 3. Reverse payload XOR
    const payloadLen = readPayloadLength(frame, 0);
    const xorRng = new Xorshift64(key.xorSeed ^ (BigInt(cipherDim) * 0x9E3779B97F4A7C15n));
    const end = Math.min(10 + payloadLen, out.length);
    for (let i = 10; i < end; i++) {
      const current = readIndexedNumber(out, i, 'helixDecryptFrame payload byte');
      out[i] = current ^ xorRng.nextByte();
    }

    // 2. Reverse phase rotation
    const rotatedPos = readU32BE(frame, 2);
    const offset =
      plainDim < 54 ? key.phaseOffsets[plainDim] ?? 0 : 0;
    const originalPos = (rotatedPos - offset) >>> 0;
    writeU32BE(out, 2, originalPos);

    // Restore original stream_id
    const originalStreamId = encodeHelixStreamId(plainDim, source);
    out[0] = (originalStreamId >> 8) & 0xFF;
    out[1] = originalStreamId & 0xFF;

    return out;

  } catch (error) {
    throw new Error(`[helixDecryptFrame] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Encrypt an entire multi-frame FlowFrame stream.
 */
export function helixEncryptStream(stream: Uint8Array, key: HelixCipherKey): Uint8Array {
  const output: Uint8Array[] = [];
  let pos = 0;

  while (pos + 10 <= stream.length) {
    const payloadLen = readPayloadLength(stream, pos);
    const frameEnd = Math.min(pos + 10 + payloadLen, stream.length);
    output.push(helixEncryptFrame(stream.subarray(pos, frameEnd), key));
    pos = frameEnd;
  }

  // Concatenate
  const totalLen = output.reduce((sum, chunk) => sum + chunk.length, 0);
  const result = new Uint8Array(totalLen);
  let offset = 0;
  for (const chunk of output) {
    result.set(chunk, offset);
    offset += chunk.length;
  }
  return result;
}

/**
 * Decrypt an entire multi-frame FlowFrame stream.
 */
export function helixDecryptStream(stream: Uint8Array, key: HelixCipherKey): Uint8Array {
  const output: Uint8Array[] = [];
  let pos = 0;

  while (pos + 10 <= stream.length) {
    const payloadLen = readPayloadLength(stream, pos);
    const frameEnd = Math.min(pos + 10 + payloadLen, stream.length);
    output.push(helixDecryptFrame(stream.subarray(pos, frameEnd), key));
    pos = frameEnd;
  }

  const totalLen = output.reduce((sum, chunk) => sum + chunk.length, 0);
  const result = new Uint8Array(totalLen);
  let offset = 0;
  for (const chunk of output) {
    result.set(chunk, offset);
    offset += chunk.length;
  }
  return result;
}

// ═══════════════════════════════════════════════════════════════════════════════
// Rotating Half-Helix Cipher
// ═══════════════════════════════════════════════════════════════════════════════

/** Samples per helical turn (closest integer to 10.5bp B-form period) */
export const HELIX_TURN_SAMPLES = 11;

/** Bit 6 flag indicating beta half in stream_id dimension byte */
const BETA_HALF_FLAG = 0x40;

/**
 * Rotating half-helix cipher key.
 *
 * Splits the 54-dimension space into two halves with independent permutations.
 * The split point rotates with helical geometry: it advances by one dimension
 * per helical turn (~11 positions).
 *
 * Keyspace: 54!^2 (~2^474) x 2^1728 phase x 2^128 XOR seeds x 2^16 stream mask
 */
export interface RotatingHelixKey {
  /** Alpha permutation (encode half). perm_alpha[plain] = cipher */
  permAlpha: Uint8Array; // 54 entries
  /** Inverse alpha. inv_alpha[cipher] = plain */
  invAlpha: Uint8Array; // 54 entries
  /** Beta permutation (decode half). perm_beta[plain] = cipher */
  permBeta: Uint8Array; // 54 entries
  /** Inverse beta. inv_beta[cipher] = plain */
  invBeta: Uint8Array; // 54 entries
  /** Per-dimension phase offsets (shared between halves) */
  phaseOffsets: Uint32Array; // 54 entries
  /** Alpha XOR seed */
  xorSeedAlpha: bigint;
  /** Beta XOR seed */
  xorSeedBeta: bigint;
  /** Stream ID XOR mask */
  streamIdMask: number;
}

/**
 * Generate a RotatingHelixKey from a 64-byte seed.
 */
export function rotatingHelixKeyFromSeed(seed: Uint8Array): RotatingHelixKey {
  try {
    if (seed.length < 64) {
      throw new Error(`Rotating helix key seed must be at least 64 bytes, got ${seed.length}`);
    }

    // Derive two independent PRNG states from seed
    let stateA = 0n;
    let stateB = 0n;
    for (let i = 0; i < seed.length; i += 8) {
      let val = 0n;
      for (let j = 0; j < 8 && i + j < seed.length; j++) {
        val |= BigInt(readIndexedNumber(seed, i + j, 'rotatingHelixKeyFromSeed seed byte')) << BigInt(j * 8);
      }
      if ((i / 8) % 2 === 0) {
        stateA ^= val;
      } else {
        stateB ^= val;
      }
    }

    const rngA = new Xorshift64(stateA);
    const rngB = new Xorshift64(stateB);

    // Fisher-Yates for alpha
    const permAlpha = new Uint8Array(54);
    for (let i = 0; i < 54; i++) permAlpha[i] = i;
    for (let i = 53; i >= 1; i--) {
      const j = Number(rngA.next() % BigInt(i + 1));
      const tmp = readIndexedNumber(permAlpha, i, 'rotatingHelixKeyFromSeed permAlpha[i]');
      const swap = readIndexedNumber(permAlpha, j, 'rotatingHelixKeyFromSeed permAlpha[j]');
      permAlpha[i] = swap;
      permAlpha[j] = tmp;
    }
    const invAlpha = new Uint8Array(54);
    for (let i = 0; i < 54; i++) invAlpha[readIndexedNumber(permAlpha, i, 'rotatingHelixKeyFromSeed invAlpha')] = i;

    // Fisher-Yates for beta
    const permBeta = new Uint8Array(54);
    for (let i = 0; i < 54; i++) permBeta[i] = i;
    for (let i = 53; i >= 1; i--) {
      const j = Number(rngB.next() % BigInt(i + 1));
      const tmp = readIndexedNumber(permBeta, i, 'rotatingHelixKeyFromSeed permBeta[i]');
      const swap = readIndexedNumber(permBeta, j, 'rotatingHelixKeyFromSeed permBeta[j]');
      permBeta[i] = swap;
      permBeta[j] = tmp;
    }
    const invBeta = new Uint8Array(54);
    for (let i = 0; i < 54; i++) invBeta[readIndexedNumber(permBeta, i, 'rotatingHelixKeyFromSeed invBeta')] = i;

    // Phase offsets from alpha RNG
    const phaseOffsets = new Uint32Array(54);
    for (let i = 0; i < 54; i++) {
      phaseOffsets[i] = Number(rngA.next() & 0xFFFFFFFFn);
    }

    return {
      permAlpha,
      invAlpha,
      permBeta,
      invBeta,
      phaseOffsets,
      xorSeedAlpha: rngA.next(),
      xorSeedBeta: rngB.next(),
      streamIdMask: Number(rngA.next() & 0xFFFFn),
    };

  } catch (error) {
    throw new Error(`[rotatingHelixKeyFromSeed] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/** Compute rotation offset for a given position */
export function rotationAt(position: number): number {
  return Math.floor(position / HELIX_TURN_SAMPLES) % 54;
}

/** Check if dimension is in the alpha (encode) half at given rotation */
export function isAlphaHalf(dimension: number, rotation: number): boolean {
  const dist = ((dimension + 54 - rotation) % 54);
  return dist < 27;
}

/**
 * Encrypt a single FlowFrame with the rotating half-helix cipher.
 */
export function rotatingHelixEncryptFrame(
  frame: Uint8Array,
  key: RotatingHelixKey
): Uint8Array {
  try {
    if (frame.length < 10) return new Uint8Array(frame);

    const out = new Uint8Array(frame);

    const streamId = (readIndexedNumber(frame, 0, 'rotatingHelixEncryptFrame streamId byte 0') << 8) |
      readIndexedNumber(frame, 1, 'rotatingHelixEncryptFrame streamId byte 1');
    const dimension = (streamId >> 8) & 0xFF;
    const source = streamId & 0xFF;
    const position = readU32BE(frame, 2);

    const rotation = rotationAt(position);
    const isBeta = !isAlphaHalf(dimension, rotation);

    // 1. Rotating dimension permutation
    const cipherDim = dimension < 54
      ? (isBeta
        ? readIndexedNumber(key.permBeta, dimension, 'rotatingHelixEncryptFrame beta permutation')
        : readIndexedNumber(key.permAlpha, dimension, 'rotatingHelixEncryptFrame alpha permutation'))
      : dimension;

    // 2. Phase rotation with helical advance
    const offset =
      dimension < 54
        ? readIndexedNumber(key.phaseOffsets, dimension, 'rotatingHelixEncryptFrame phase offset')
        : 0;
    const rotatedPos = (position + offset + rotation * 7920) >>> 0;
    writeU32BE(out, 2, rotatedPos);

    // 3. Payload XOR with rotation-aware seed
    const payloadLen = readPayloadLength(frame, 0);
    const baseSeed = isBeta ? key.xorSeedBeta : key.xorSeedAlpha;
    const xorSeed = baseSeed
      ^ (BigInt(cipherDim) * 0x9E3779B97F4A7C15n)
      ^ (BigInt(rotation) * 0x517CC1B727220A95n);
    const xorRng = new Xorshift64(xorSeed & 0xFFFFFFFFFFFFFFFFn);
    const end = Math.min(10 + payloadLen, out.length);
    for (let i = 10; i < end; i++) {
      const current = readIndexedNumber(out, i, 'rotatingHelixEncryptFrame payload byte');
      out[i] = current ^ xorRng.nextByte();
    }

    // 4. Stream ID scramble with half-indicator
    const dimWithFlag = cipherDim | (isBeta ? BETA_HALF_FLAG : 0);
    const newStreamId = ((dimWithFlag << 8) | source) ^ key.streamIdMask;
    out[0] = (newStreamId >> 8) & 0xFF;
    out[1] = newStreamId & 0xFF;

    return out;

  } catch (error) {
    throw new Error(`[rotatingHelixEncryptFrame] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Decrypt a single FlowFrame with the rotating half-helix cipher.
 */
export function rotatingHelixDecryptFrame(
  frame: Uint8Array,
  key: RotatingHelixKey
): Uint8Array {
  try {
    if (frame.length < 10) return new Uint8Array(frame);

    const out = new Uint8Array(frame);

    // 4. Reverse stream ID scramble
    const scrambledId = (readIndexedNumber(frame, 0, 'rotatingHelixDecryptFrame streamId byte 0') << 8) |
      readIndexedNumber(frame, 1, 'rotatingHelixDecryptFrame streamId byte 1');
    const unscrambledId = scrambledId ^ key.streamIdMask;
    const dimWithFlag = (unscrambledId >> 8) & 0xFF;
    const isBeta = (dimWithFlag & BETA_HALF_FLAG) !== 0;
    const cipherDim = dimWithFlag & 0x3F;
    const source = unscrambledId & 0xFF;

    // 1. Reverse dimension permutation
    const plainDim = isBeta
      ? readIndexedNumber(key.invBeta, cipherDim, 'rotatingHelixDecryptFrame inverse beta')
      : readIndexedNumber(key.invAlpha, cipherDim, 'rotatingHelixDecryptFrame inverse alpha');

    // 2. Reverse phase rotation -- recover original position
    const encryptedPos = readU32BE(frame, 2);
    const offset =
      plainDim < 54
        ? readIndexedNumber(key.phaseOffsets, plainDim, 'rotatingHelixDecryptFrame phase offset')
        : 0;

    // O(1) rotation recovery using modular inverse
    const base = (encryptedPos - offset) >>> 0;
    const b = Math.floor(base / 11);
    const rotCandidate = (b * 37) % 54;
    const candidatePos = (base - rotCandidate * 7920) >>> 0;

    let foundRotation: number;
    let foundPos: number;

    if (rotationAt(candidatePos) === rotCandidate) {
      foundRotation = rotCandidate;
      foundPos = candidatePos;
    } else {
      // Fallback for u32 wrapping edge cases
      foundRotation = 0;
      foundPos = encryptedPos;
      for (let rot = 0; rot < 54; rot++) {
        const cp = (base - rot * 7920) >>> 0;
        if (rotationAt(cp) === rot) {
          foundRotation = rot;
          foundPos = cp;
          break;
        }
      }
    }

    // 3. Reverse payload XOR
    const payloadLen = readPayloadLength(frame, 0);
    const baseSeed = isBeta ? key.xorSeedBeta : key.xorSeedAlpha;
    const xorSeed = baseSeed
      ^ (BigInt(cipherDim) * 0x9E3779B97F4A7C15n)
      ^ (BigInt(foundRotation) * 0x517CC1B727220A95n);
    const xorRng = new Xorshift64(xorSeed & 0xFFFFFFFFFFFFFFFFn);
    const end = Math.min(10 + payloadLen, out.length);
    for (let i = 10; i < end; i++) {
      const current = readIndexedNumber(out, i, 'rotatingHelixDecryptFrame payload byte');
      out[i] = current ^ xorRng.nextByte();
    }

    // Restore original stream_id and position
    const originalStreamId = encodeHelixStreamId(plainDim, source);
    out[0] = (originalStreamId >> 8) & 0xFF;
    out[1] = originalStreamId & 0xFF;
    writeU32BE(out, 2, foundPos);

    return out;

  } catch (error) {
    throw new Error(`[rotatingHelixDecryptFrame] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Encrypt an entire multi-frame stream with the rotating half-helix cipher.
 */
export function rotatingHelixEncryptStream(
  stream: Uint8Array,
  key: RotatingHelixKey
): Uint8Array {
  const output: Uint8Array[] = [];
  let pos = 0;

  while (pos + 10 <= stream.length) {
    const payloadLen = readPayloadLength(stream, pos);
    const frameEnd = Math.min(pos + 10 + payloadLen, stream.length);
    output.push(rotatingHelixEncryptFrame(stream.subarray(pos, frameEnd), key));
    pos = frameEnd;
  }

  const totalLen = output.reduce((sum, chunk) => sum + chunk.length, 0);
  const result = new Uint8Array(totalLen);
  let offset = 0;
  for (const chunk of output) {
    result.set(chunk, offset);
    offset += chunk.length;
  }
  return result;
}

/**
 * Decrypt an entire multi-frame stream with the rotating half-helix cipher.
 */
export function rotatingHelixDecryptStream(
  stream: Uint8Array,
  key: RotatingHelixKey
): Uint8Array {
  const output: Uint8Array[] = [];
  let pos = 0;

  while (pos + 10 <= stream.length) {
    const payloadLen = readPayloadLength(stream, pos);
    const frameEnd = Math.min(pos + 10 + payloadLen, stream.length);
    output.push(rotatingHelixDecryptFrame(stream.subarray(pos, frameEnd), key));
    pos = frameEnd;
  }

  const totalLen = output.reduce((sum, chunk) => sum + chunk.length, 0);
  const result = new Uint8Array(totalLen);
  let offset = 0;
  for (const chunk of output) {
    result.set(chunk, offset);
    offset += chunk.length;
  }
  return result;
}

// ═══════════════════════════════════════════════════════════════════════════════
// Convenience: derive key from passphrase via SHA-256
// ═══════════════════════════════════════════════════════════════════════════════

/**
 * Derive a 64-byte seed from a passphrase using SHA-256 (double-hash).
 * For production, prefer HKDF or Argon2.
 */
export async function seedFromPassphrase(passphrase: string): Promise<Uint8Array> {
  const encoder = new TextEncoder();
  const data = encoder.encode(passphrase);
  const hash1 = new Uint8Array(await crypto.subtle.digest('SHA-256', data));
  const hash2 = new Uint8Array(await crypto.subtle.digest('SHA-256', hash1));
  const seed = new Uint8Array(64);
  seed.set(hash1, 0);
  seed.set(hash2, 32);
  return seed;
}

const BITWISE_MODULE_SPECIFIER = '@a0n/bitwise' as string;

async function loadBitwiseModule(): Promise<any> {
  return import(BITWISE_MODULE_SPECIFIER);
}

/**
 * Handles the gnosis syzygy Smush workflow.
 */
export async function syzygySmush(payload: Uint8Array, keyPhrase: string): Promise<Uint8Array> {
  try {
    const mod = await loadBitwiseModule();
    if (mod.syzygySmush) return mod.syzygySmush(payload, keyPhrase);
    if (mod.loadBitwiseClient) return (await mod.loadBitwiseClient()).syzygySmush(payload, keyPhrase);
    return (mod as any).default.syzygySmush(payload, keyPhrase);

  } catch (error) {
    throw new Error(`[syzygySmush] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Handles the gnosis syzygy Unsmush workflow.
 */
export async function syzygyUnsmush(payload: Uint8Array, keyPhrase: string, originalLength: number): Promise<Uint8Array> {
  try {
    const mod = await loadBitwiseModule();
    if (mod.syzygyUnsmush) return mod.syzygyUnsmush(payload, keyPhrase, originalLength);
    if (mod.loadBitwiseClient) return (await mod.loadBitwiseClient()).syzygyUnsmush(payload, keyPhrase, originalLength);
    return (mod as any).default.syzygyUnsmush(payload, keyPhrase, originalLength);

  } catch (error) {
    throw new Error(`[syzygyUnsmush] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Handles the gnosis syzygy Fold Binary workflow.
 */
export async function syzygyFoldBinary(payload: Uint8Array): Promise<Uint8Array> {
  try {
    const mod = await loadBitwiseModule();
    if (mod.syzygyFoldBinary) return mod.syzygyFoldBinary(payload);
    if (mod.loadBitwiseClient) return (await mod.loadBitwiseClient()).syzygyFoldBinary(payload);
    return (mod as any).default.syzygyFoldBinary(payload);

  } catch (error) {
    throw new Error(`[syzygyFoldBinary] ${error instanceof Error ? error.message : String(error)}`);
  }
}



/**
 * Handles the gnosis syzygy Unfold Binary workflow.
 */
export async function syzygyUnfoldBinary(payload: Uint8Array, originalLength: number): Promise<Uint8Array> {
  try {
    const mod = await loadBitwiseModule();
    if (mod.syzygyUnfoldBinary) return mod.syzygyUnfoldBinary(payload, originalLength);
    if (mod.loadBitwiseClient) return (await mod.loadBitwiseClient()).syzygyUnfoldBinary(payload, originalLength);
    return (mod as any).default.syzygyUnfoldBinary(payload, originalLength);

  } catch (error) {
    throw new Error(`[syzygyUnfoldBinary] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Encrypt a FlowFrame using the full Lacey cipher variant:
 * 1. Structural rotating half-helix cipher (steganographic confusion).
 * 2. Syzygy smush (ChaCha20-Poly1305 AEAD via WASM).
 */
export async function laceyEncryptFrame(
  frame: Uint8Array,
  key: RotatingHelixKey,
  keyPhrase: string
): Promise<Uint8Array> {
  try {
    const structural = rotatingHelixEncryptFrame(frame, key);
    const mod = await loadBitwiseModule();
    if (mod.syzygySmush) return mod.syzygySmush(structural, keyPhrase);
    if (mod.loadBitwiseClient) return (await mod.loadBitwiseClient()).syzygySmush(structural, keyPhrase);
    return (mod as any).default.syzygySmush(structural, keyPhrase);

  } catch (error) {
    throw new Error(`[laceyEncryptFrame] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Decrypt a FlowFrame using the full Lacey cipher variant:
 * 1. Syzygy unsmush (ChaCha20-Poly1305 AEAD via WASM).
 * 2. Structural rotating half-helix cipher (un-rotate and un-permute).
 */
export async function laceyDecryptFrame(
  encrypted: Uint8Array,
  key: RotatingHelixKey,
  keyPhrase: string,
  originalLength: number
): Promise<Uint8Array> {
  try {
    const mod = await loadBitwiseModule();
    let structural: Uint8Array;
    if (mod.syzygyUnsmush) {
      structural = await mod.syzygyUnsmush(encrypted, keyPhrase, originalLength);
    } else if (mod.loadBitwiseClient) {
      structural = await (await mod.loadBitwiseClient()).syzygyUnsmush(encrypted, keyPhrase, originalLength);
    } else {
      structural = await (mod as any).default.syzygyUnsmush(encrypted, keyPhrase, originalLength);
    }
    return rotatingHelixDecryptFrame(structural, key);

  } catch (error) {
    throw new Error(`[laceyDecryptFrame] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Encrypt an entire multi-frame stream using the full Lacey cipher variant.
 */
export async function laceyEncryptStream(
  stream: Uint8Array,
  key: RotatingHelixKey,
  keyPhrase: string
): Promise<Uint8Array> {
  try {
    const structural = rotatingHelixEncryptStream(stream, key);
    const mod = await loadBitwiseModule();
    if (mod.syzygySmush) return mod.syzygySmush(structural, keyPhrase);
    if (mod.loadBitwiseClient) return (await mod.loadBitwiseClient()).syzygySmush(structural, keyPhrase);
    return (mod as any).default.syzygySmush(structural, keyPhrase);

  } catch (error) {
    throw new Error(`[laceyEncryptStream] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Decrypt an entire multi-frame stream using the full Lacey cipher variant.
 */
export async function laceyDecryptStream(
  encrypted: Uint8Array,
  key: RotatingHelixKey,
  keyPhrase: string,
  originalLength: number
): Promise<Uint8Array> {
  try {
    const mod = await loadBitwiseModule();
    let structural: Uint8Array;
    if (mod.syzygyUnsmush) {
      structural = await mod.syzygyUnsmush(encrypted, keyPhrase, originalLength);
    } else if (mod.loadBitwiseClient) {
      structural = await (await mod.loadBitwiseClient()).syzygyUnsmush(encrypted, keyPhrase, originalLength);
    } else {
      structural = await (mod as any).default.syzygyUnsmush(encrypted, keyPhrase, originalLength);
    }
    return rotatingHelixDecryptStream(structural, key);

  } catch (error) {
    throw new Error(`[laceyDecryptStream] ${error instanceof Error ? error.message : String(error)}`);
  }
}
