// @a0n/gnosis/src/causal-structure.ts
/**
 * causal-structure.ts -- Spacetime and causality from topology ordering
 *
 * The identification:
 *   - DAG edges = causal relations (A → B means A is in B's causal past)
 *   - Topological depth = proper time (longer paths = more elapsed time)
 *   - FORK = light cone expansion (future branches)
 *   - FOLD = causal diamond closure (past converges)
 *   - Complement distribution over reachable nodes = causal horizon
 *   - Unreachable nodes = spacelike separated
 *
 * The formal ledger entry: every topology maps to a causal structure.
 */

import {
  type VoidBoundary,
  createVoidBoundary,
  updateVoidBoundary,
  complementDistribution,
} from './void';

// ============================================================================
// Causal relations from DAG structure
// ============================================================================

export interface CausalNode {
  id: string;
  depth: number;
  /** Indices of nodes in causal past (ancestors) */
  past: Set<number>;
  /** Indices of nodes in causal future (descendants) */
  future: Set<number>;
}

export interface CausalStructure {
  nodes: CausalNode[];
  nodeIndex: Map<string, number>;
  /** Adjacency list: edges[i] = indices of direct successors */
  edges: number[][];
  /** Maximum depth (proper time of longest causal chain) */
  maxDepth: number;
}

/**
 * Build causal structure from a DAG (adjacency list of node IDs).
 */
export function buildCausalStructure(
  nodeIds: string[],
  edgeList: [string, string][]
): CausalStructure {
  try {
    const nodeIndex = new Map<string, number>();
    nodeIds.forEach((id, i) => nodeIndex.set(id, i));

    const n = nodeIds.length;
    const edges: number[][] = Array.from({ length: n }, () => []);

    for (const [src, tgt] of edgeList) {
      const si = nodeIndex.get(src);
      const ti = nodeIndex.get(tgt);
      if (si !== undefined && ti !== undefined) {
        edges[si].push(ti);
      }
    }

    // Compute depths via topological sort
    const inDegree = new Array(n).fill(0);
    for (const successors of edges) {
      for (const t of successors) inDegree[t]++;
    }
    const depth = new Array(n).fill(0);
    const queue: number[] = [];
    for (let i = 0; i < n; i++) {
      if (inDegree[i] === 0) queue.push(i);
    }
    while (queue.length > 0) {
      const u = queue.shift()!;
      for (const v of edges[u]) {
        depth[v] = Math.max(depth[v], depth[u] + 1);
        inDegree[v]--;
        if (inDegree[v] === 0) queue.push(v);
      }
    }

    // Compute transitive closure (past/future)
    const past: Set<number>[] = Array.from({ length: n }, () => new Set());
    const future: Set<number>[] = Array.from({ length: n }, () => new Set());

    // Process in topological order (by depth)
    const sorted = Array.from({ length: n }, (_, i) => i).sort(
      (a, b) => depth[a] - depth[b]
    );
    for (const u of sorted) {
      for (const v of edges[u]) {
        future[u].add(v);
        past[v].add(u);
        // Transitive: u's past is also in v's past
        for (const p of past[u]) {
          past[v].add(p);
          future[p].add(v);
        }
      }
    }

    const nodes: CausalNode[] = nodeIds.map((id, i) => ({
      id,
      depth: depth[i],
      past: past[i],
      future: future[i],
    }));

    return {
      nodes,
      nodeIndex,
      edges,
      maxDepth: Math.max(...depth, 0),
    };

  } catch (error) {
    throw new Error(`[buildCausalStructure] ${error instanceof Error ? error.message : String(error)}`);
  }
}

// ============================================================================
// Causal relations
// ============================================================================

export type CausalRelation = 'past' | 'future' | 'spacelike' | 'identical';

/**
 * Determine the causal relation between two nodes.
 */
export function causalRelation(
  cs: CausalStructure,
  a: number,
  b: number
): CausalRelation {
  try {
    if (a === b) return 'identical';
    if (cs.nodes[a].future.has(b)) return 'future';
    if (cs.nodes[a].past.has(b)) return 'past';
    return 'spacelike';

  } catch (error) {
    throw new Error(`[causalRelation] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Light cone of a node: all causally connected nodes.
 */
export function lightCone(
  cs: CausalStructure,
  nodeIdx: number
): {
  past: number[];
  future: number[];
  spacelike: number[];
} {
  try {
    const node = cs.nodes[nodeIdx];
    const past = [...node.past];
    const future = [...node.future];
    const spacelike: number[] = [];
    for (let i = 0; i < cs.nodes.length; i++) {
      if (i !== nodeIdx && !node.past.has(i) && !node.future.has(i)) {
        spacelike.push(i);
      }
    }
    return { past, future, spacelike };

  } catch (error) {
    throw new Error(`[lightCone] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Causal diamond: intersection of A's future cone and B's past cone.
 */
export function causalDiamond(
  cs: CausalStructure,
  a: number,
  b: number
): number[] {
  const aFuture = cs.nodes[a].future;
  const bPast = cs.nodes[b].past;
  return [...aFuture].filter((i) => bPast.has(i));
}

/**
 * Proper time between two causally connected nodes.
 * This is the depth difference (longest path length).
 */
export function properTime(
  cs: CausalStructure,
  a: number,
  b: number
): number | null {
  try {
    if (!cs.nodes[a].future.has(b) && a !== b) return null;
    return cs.nodes[b].depth - cs.nodes[a].depth;

  } catch (error) {
    throw new Error(`[properTime] ${error instanceof Error ? error.message : String(error)}`);
  }
}

// ============================================================================
// Causal VoidBoundary -- void accumulates at visited causal events
// ============================================================================

/**
 * Convert causal structure to VoidBoundary.
 * Each node is a dimension. Deeper nodes (more proper time elapsed)
 * accumulate more void. The complement distribution then peaks at
 * "fresh" nodes -- those not yet in the causal past.
 */
export function causalToVoidBoundary(cs: CausalStructure): VoidBoundary {
  const boundary = createVoidBoundary(cs.nodes.length);
  for (let i = 0; i < cs.nodes.length; i++) {
    // Void = size of causal past (more history = more void)
    updateVoidBoundary(boundary, i, cs.nodes[i].past.size);
  }
  return boundary;
}

/**
 * Causal horizon: the complement distribution over a causal VoidBoundary
 * peaks at nodes with the smallest causal past -- the "freshest" events.
 */
export function causalHorizon(
  cs: CausalStructure,
  eta: number = 3.0
): number[] {
  const boundary = causalToVoidBoundary(cs);
  return complementDistribution(boundary, eta);
}
