// @a0n/gnosis/src/clock-distributed.ts
/**
 * Distributed Clock — time across AeonPIDs.
 *
 * Every AeonPID is a process with its own clock. When two PIDs communicate,
 * propagation delay means they disagree about "now." The phase difference
 * between them IS the network distance, measured in time.
 *
 * This module maps relativistic spacetime concepts to distributed systems:
 *
 *   Light cone    → message delivery horizon
 *   Lookback time → propagation delay
 *   Simultaneity  → event ordering ambiguity
 *   Solar phase   → process activity phase (active/idle/cold)
 *   Clock mesh    → network temporal topology
 *
 * The speed of light through fiber is ~0.67c. Two edge nodes 10,000 km
 * apart have ~50ms of irreducible latency. That's their "spacelike
 * separation" -- they cannot agree on "now" within that window.
 *
 * @packageDocumentation
 */

import {
  PICOLORENZO_MS,
  PICOLORENZO_SECONDS,
  LIGHT_PICOLORENZO_LY,
  formatGnosticTime,
  toPicolorenzos,
} from './math';

// ─── Constants ──────────────────────────────────────────────────────

/** Speed of light in fiber optic cable, as fraction of c (~0.67). */
export const FIBER_SPEED_FRACTION_C = 0.67;

/** Speed of light in fiber, in km/ms. c = 299,792 km/s ≈ 299.8 km/ms. */
const C_KM_PER_MS = 299.792;
const FIBER_KM_PER_MS = C_KM_PER_MS * FIBER_SPEED_FRACTION_C;

/** Convert milliseconds to picolorenzos. */
function msToPLo(ms: number): number {
  return ms / PICOLORENZO_MS;
}

/** Convert picolorenzos to milliseconds. */
function pLoToMs(pLo: number): number {
  return pLo * PICOLORENZO_MS;
}

// ─── Process Clock ──────────────────────────────────────────────────

/** Activity phase of a distributed process. */
export type ProcessPhase =
  | 'active'      // processing requests (lit phase)
  | 'idle'        // alive but no work (twilight)
  | 'cold'        // not started / cold-starting (dark phase)
  | 'unreachable'; // no response (beyond horizon)

/**
 * A distributed process with a clock.
 * Every AeonPID has one of these.
 */
export interface ProcessClock {
  readonly id: string;
  readonly label: string;
  /** Geographic location of the edge node. */
  readonly location: {
    readonly lat: number;
    readonly lon: number;
    readonly region: string; // e.g., "us-west-1", "eu-central-1"
  };
  /** Current phase. */
  readonly phase: ProcessPhase;
  /** Local clock reading (epoch ms). */
  readonly localEpochMs: number;
  /** Last heartbeat received (epoch ms). null if never heard from. */
  readonly lastHeartbeatMs: number | null;
  /** Request rate (requests/second). 0 if idle/cold. */
  readonly requestRate: number;
}

// ─── Clock Distance ─────────────────────────────────────────────────

/**
 * Distance between two clocks, measured in time.
 *
 * This is the fundamental measurement: not "how far apart" but
 * "how long does it take to communicate." Distance IS time.
 */
export interface ClockDistance {
  readonly fromId: string;
  readonly toId: string;
  /** Great-circle distance in km. */
  readonly spatialDistanceKm: number;
  /** One-way propagation delay through fiber, in milliseconds. */
  readonly propagationDelayMs: number;
  /** One-way propagation delay in picolorenzos. */
  readonly propagationDelayPLo: number;
  /** Round-trip time in milliseconds. */
  readonly roundTripMs: number;
  /** Round-trip time in picolorenzos. */
  readonly roundTripPLo: number;
  /**
   * Simultaneity window: events within this time window at the two
   * locations cannot be causally ordered. Neither happened "before"
   * the other. This is the irreducible uncertainty in "now."
   */
  readonly simultaneityWindowMs: number;
  /** In gnostic time units. */
  readonly simultaneityGnostic: string;
  /**
   * Can these two clocks agree on event ordering within the given
   * tolerance? If the tolerance is smaller than the propagation delay,
   * they cannot -- Lamport ordering is required.
   */
  readonly requiresLogicalClock: boolean;
}

/**
 * Haversine great-circle distance between two lat/lon points, in km.
 */
function haversineKm(
  lat1: number, lon1: number,
  lat2: number, lon2: number
): number {
  const R = 6371; // Earth radius in km
  const dLat = (lat2 - lat1) * Math.PI / 180;
  const dLon = (lon2 - lon1) * Math.PI / 180;
  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos(lat1 * Math.PI / 180) * Math.cos(lat2 * Math.PI / 180) *
    Math.sin(dLon / 2) * Math.sin(dLon / 2);
  return R * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}

/**
 * Compute the clock distance between two process clocks.
 *
 * The distance is the propagation delay through fiber optic cable
 * along the great-circle path. The simultaneity window is the
 * round-trip time -- events within this window at the two nodes
 * cannot be causally ordered by physical clocks alone.
 */
export function clockDistance(
  from: ProcessClock,
  to: ProcessClock,
  /** Ordering tolerance in ms. Default: 1ms. */
  toleranceMs?: number
): ClockDistance {
  const distKm = haversineKm(
    from.location.lat, from.location.lon,
    to.location.lat, to.location.lon
  );

  // Fiber path is ~1.4x great-circle distance (routing overhead)
  const fiberPathKm = distKm * 1.4;
  const propagationMs = fiberPathKm / FIBER_KM_PER_MS;
  const roundTripMs = propagationMs * 2;
  const propagationPLo = msToPLo(propagationMs);
  const roundTripPLo = msToPLo(roundTripMs);
  const tolerance = toleranceMs ?? 1;

  return {
    fromId: from.id,
    toId: to.id,
    spatialDistanceKm: distKm,
    propagationDelayMs: propagationMs,
    propagationDelayPLo: propagationPLo,
    roundTripMs,
    roundTripPLo,
    simultaneityWindowMs: roundTripMs,
    simultaneityGnostic: formatGnosticTime(roundTripPLo),
    requiresLogicalClock: propagationMs > tolerance,
  };
}

// ─── Distributed Clock Mesh ─────────────────────────────────────────

/**
 * An edge in the distributed clock network.
 */
export interface ClockNetworkEdge {
  readonly fromId: string;
  readonly toId: string;
  readonly distance: ClockDistance;
  /**
   * Causal classification (borrowed from spacetime):
   * - 'PROCESS': within tolerance, physical clocks sufficient
   * - 'VENT': beyond tolerance, requires logical clock (Lamport/vector)
   */
  readonly edgeType: 'PROCESS' | 'VENT';
}

/**
 * The temporal topology of a distributed system.
 *
 * A mesh of process clocks connected by propagation delays.
 * Shows which processes can agree on "now" (PROCESS edges) and
 * which require logical ordering (VENT edges).
 */
export interface DistributedClockMesh {
  readonly source: 'GnosisDistributedClockMesh';
  readonly clocks: readonly ProcessClock[];
  readonly edges: readonly ClockNetworkEdge[];
  /** Fraction of edges where physical clock ordering is sufficient. */
  readonly physicalOrderingFraction: number;
  /** Maximum propagation delay in the network (diameter), in ms. */
  readonly networkDiameterMs: number;
  /** Maximum propagation delay in picolorenzos. */
  readonly networkDiameterPLo: number;
  /** Average propagation delay across all edges, in ms. */
  readonly meanPropagationMs: number;
  /** Number of active (lit) processes. */
  readonly activeCount: number;
  /** Number of idle (twilight) processes. */
  readonly idleCount: number;
  /** Number of cold/unreachable (dark) processes. */
  readonly darkCount: number;
  /** Lit fraction (active / total). */
  readonly litFraction: number;
}

/**
 * Build the temporal topology of a distributed AeonPID network.
 *
 * Every pair of process clocks gets an edge with propagation delay
 * and causal classification. The result shows the causal structure
 * of the distributed system -- which nodes can physically agree on
 * event ordering, and which need Lamport clocks.
 */
export function createDistributedClockMesh(
  clocks: readonly ProcessClock[],
  toleranceMs?: number
): DistributedClockMesh {
  const edges: ClockNetworkEdge[] = [];
  let maxPropMs = 0;
  let totalPropMs = 0;
  let physicalCount = 0;

  for (let i = 0; i < clocks.length; i++) {
    for (let j = i + 1; j < clocks.length; j++) {
      const dist = clockDistance(clocks[i]!, clocks[j]!, toleranceMs);
      const edgeType: 'PROCESS' | 'VENT' = dist.requiresLogicalClock ? 'VENT' : 'PROCESS';
      if (!dist.requiresLogicalClock) physicalCount++;
      if (dist.propagationDelayMs > maxPropMs) maxPropMs = dist.propagationDelayMs;
      totalPropMs += dist.propagationDelayMs;

      edges.push({
        fromId: clocks[i]!.id,
        toId: clocks[j]!.id,
        distance: dist,
        edgeType,
      });
    }
  }

  const activeCount = clocks.filter(c => c.phase === 'active').length;
  const idleCount = clocks.filter(c => c.phase === 'idle').length;
  const darkCount = clocks.filter(c => c.phase === 'cold' || c.phase === 'unreachable').length;

  return {
    source: 'GnosisDistributedClockMesh',
    clocks,
    edges,
    physicalOrderingFraction: edges.length > 0 ? physicalCount / edges.length : 1,
    networkDiameterMs: maxPropMs,
    networkDiameterPLo: msToPLo(maxPropMs),
    meanPropagationMs: edges.length > 0 ? totalPropMs / edges.length : 0,
    activeCount,
    idleCount,
    darkCount,
    litFraction: clocks.length > 0 ? activeCount / clocks.length : 0,
  };
}

// ─── Lamport Clock ──────────────────────────────────────────────────

/**
 * A Lamport logical clock for causal ordering.
 *
 * When physical clocks can't agree on "now" (VENT edges),
 * Lamport clocks provide causal ordering: if A → B then L(A) < L(B).
 *
 * Each process maintains a counter. On local event: increment.
 * On send: attach counter. On receive: max(local, received) + 1.
 */
export interface LamportClock {
  readonly processId: string;
  readonly counter: number;
}

export interface LamportEvent {
  readonly processId: string;
  readonly counter: number;
  readonly epochMs: number;
  readonly kind: 'local' | 'send' | 'receive';
  readonly label: string;
}

/**
 * Create a new Lamport clock for a process.
 */
export function createLamportClock(processId: string): LamportClock {
  return { processId, counter: 0 };
}

/**
 * Advance a Lamport clock on a local event.
 */
export function lamportLocalEvent(
  clock: LamportClock,
  label: string
): { readonly clock: LamportClock; readonly event: LamportEvent } {
  const next = { processId: clock.processId, counter: clock.counter + 1 };
  return {
    clock: next,
    event: {
      processId: clock.processId,
      counter: next.counter,
      epochMs: Date.now(),
      kind: 'local',
      label,
    },
  };
}

/**
 * Advance a Lamport clock on sending a message.
 */
export function lamportSend(
  clock: LamportClock,
  label: string
): { readonly clock: LamportClock; readonly event: LamportEvent } {
  const next = { processId: clock.processId, counter: clock.counter + 1 };
  return {
    clock: next,
    event: {
      processId: clock.processId,
      counter: next.counter,
      epochMs: Date.now(),
      kind: 'send',
      label,
    },
  };
}

/**
 * Advance a Lamport clock on receiving a message.
 * Takes the max of local and received counter, then increments.
 */
export function lamportReceive(
  clock: LamportClock,
  receivedCounter: number,
  label: string
): { readonly clock: LamportClock; readonly event: LamportEvent } {
  const merged = Math.max(clock.counter, receivedCounter) + 1;
  const next = { processId: clock.processId, counter: merged };
  return {
    clock: next,
    event: {
      processId: clock.processId,
      counter: next.counter,
      epochMs: Date.now(),
      kind: 'receive',
      label,
    },
  };
}

/**
 * Can we determine causal ordering between two Lamport events?
 *
 * If they're from the same process, counter order = causal order.
 * If from different processes, we can only say A → B if L(A) < L(B)
 * AND there's a message chain connecting them. Without the chain,
 * equal or close counters mean "concurrent" -- the distributed
 * equivalent of spacelike separation.
 */
export function lamportHappensBefore(a: LamportEvent, b: LamportEvent): boolean | null {
  if (a.processId === b.processId) {
    return a.counter < b.counter;
  }
  // Cross-process: counter ordering is necessary but not sufficient.
  // Return null for "concurrent / unknown" (spacelike).
  if (a.counter < b.counter && a.kind === 'send' && b.kind === 'receive') {
    return true;
  }
  return null; // concurrent -- cannot determine ordering
}
