// @a0n/gnosis/src/clock.ts
/**
 * Clock — time as experienced, not observed.
 *
 * A Clock is a worldline with a counter. It doesn't see the universe.
 * It sees its own accumulated ticks. Every feature we built from the
 * atlas perspective (intervals, light cones, horizons) has a dual
 * from the clock perspective:
 *
 *   Atlas says: "these two events are spacelike separated"
 *   Clock says: "I have ticked 493 picolorenzos since I last saw that star"
 *
 *   Atlas says: "gravitational dilation factor is 0.9487 at 10 r_s"
 *   Clock says: "I am ticking at 94.87% speed"
 *
 *   Atlas says: "the sun is at altitude 45 degrees"
 *   Clock says: "I am in my lit phase"
 *
 * A Clock unifies worldlines, dilation, and solar phase into a single
 * object that tracks what *this particular clock* experiences at each
 * point along its worldline.
 *
 * @packageDocumentation
 */

import type { WorldlineTrace, WorldlinePoint } from './worldline';
import { worldlineFromPositions, stationaryWorldline, circularOrbitWorldline } from './worldline';
import type { GravitationalBody } from './gravitational-dilation';
import { gravitationalDilationFactor, combinedDilation } from './gravitational-dilation';
import type { SolarObserver, SolarPosition, SolarPhase, TwilightZone, SolarHalfDay } from './math';
import {
  solarPosition,
  solarPhase,
  solarPhaseContinuous,
  solarHalfDay,
  classifyTwilightZone,
  toPicolorenzos,
  fromPicolorenzos,
  formatGnosticTime,
  decomposeGnosticTime,
  PICOLORENZO_MS,
  PICOLORENZO_SECONDS,
} from './math';

// ─── Clock Reading ──────────────────────────────────────────────────

/**
 * A single reading from a clock at one point along its worldline.
 * Everything a clock knows about itself at this instant.
 */
export interface ClockReading {
  /** Coordinate time (picolorenzos since Unix epoch). What a distant observer's clock says. */
  readonly coordinateTimePLo: number;
  /** Proper time accumulated along this clock's worldline (picolorenzos). What THIS clock says. */
  readonly properTimePLo: number;
  /** Current tick rate: dτ/dt. 1.0 in flat spacetime, < 1 near mass or at speed. */
  readonly tickRate: number;
  /** Position in light-years. Where the clock is. */
  readonly x: number;
  readonly y: number;
  readonly z: number;
  /** Speed as fraction of c (|v|/c). 0 for stationary clocks. */
  readonly velocityFractionC: number;
  /** Gravitational dilation factor at this position (1.0 if no nearby mass). */
  readonly gravitationalTickRate: number;
  /** Kinematic dilation factor from velocity (1/γ). 1.0 if stationary. */
  readonly kinematicTickRate: number;
  /** Solar phase at this position. null if no observer location specified. */
  readonly solarPhase: SolarPhase | null;
  /** Continuous solar phase [0, 1]. null if no observer location. */
  readonly solarPhaseContinuous: number | null;
  /** Twilight zone. null if no observer location. */
  readonly twilightZone: TwilightZone | null;
  /** Solar altitude in degrees. null if no observer location. */
  readonly solarAltitudeDeg: number | null;
  /** Gnostic time label for the proper time reading. */
  readonly gnosticTime: string;
  /** How far this clock has drifted from coordinate time (proper - coordinate, in pLo). */
  readonly driftPLo: number;
}

// ─── Clock ──────────────────────────────────────────────────────────

/**
 * A Clock: a worldline with a counter, dilation context, and solar phase.
 *
 * This is what time looks like from the *inside*. Not "what happened
 * at these coordinates" but "what did this clock experience."
 */
export interface Clock {
  readonly source: 'GnosisClock';
  readonly id: string;
  readonly label: string;
  /** The underlying worldline. */
  readonly worldline: WorldlineTrace;
  /** A reading at every worldline point. */
  readonly readings: readonly ClockReading[];
  /** Total proper time elapsed (what the clock says). */
  readonly totalProperTimePLo: number;
  /** Total coordinate time elapsed (what a distant observer says). */
  readonly totalCoordinateTimePLo: number;
  /** Average tick rate across the worldline. */
  readonly averageTickRate: number;
  /** Total drift from coordinate time (positive = clock ran fast, negative = ran slow). */
  readonly totalDriftPLo: number;
  /** The solar half-day at the clock's position (if observer specified). */
  readonly solarHalfDay: SolarHalfDay | null;
}

// ─── Clock Comparison ───────────────────────────────────────────────

/**
 * What you learn when two clocks that started together are compared.
 */
export interface ClockComparison {
  readonly source: 'GnosisClockComparison';
  readonly clockA: Clock;
  readonly clockB: Clock;
  /** Which clock is ahead (shows more elapsed time). */
  readonly aheadClockId: string;
  /** Absolute difference in proper time (picolorenzos). */
  readonly properTimeDifferencePLo: number;
  /** Difference in years. */
  readonly properTimeDifferenceYears: number;
  /** Ratio of proper times (A/B). > 1 means A ticked more. */
  readonly properTimeRatio: number;
  /** The two clocks' readings at each shared coordinate time. */
  readonly syncPoints: readonly ClockSyncPoint[];
}

/**
 * A moment where two clocks are compared at the same coordinate time.
 */
export interface ClockSyncPoint {
  readonly coordinateTimePLo: number;
  readonly clockAProperTimePLo: number;
  readonly clockBProperTimePLo: number;
  readonly driftPLo: number; // A.proper - B.proper
  readonly clockATickRate: number;
  readonly clockBTickRate: number;
}

// ─── Clock Construction ─────────────────────────────────────────────

export interface CreateClockInput {
  readonly id: string;
  readonly label: string;
  readonly worldline: WorldlineTrace;
  /** Gravitational bodies affecting this clock's tick rate. */
  readonly gravitationalBodies?: readonly GravitationalBody[];
  /** Observer location for solar phase (if this clock is on a planet surface). */
  readonly observer?: SolarObserver;
}

/**
 * Create a Clock from a worldline and its environment.
 *
 * For each point on the worldline, computes:
 * - Gravitational tick rate from nearby masses
 * - Kinematic tick rate from velocity (estimated from consecutive positions)
 * - Combined tick rate (product of gravitational and kinematic)
 * - Solar phase and twilight zone (if observer specified)
 * - Drift from coordinate time
 */
export function createClock(input: CreateClockInput): Clock {
  const { id, label, worldline, gravitationalBodies, observer } = input;
  const points = worldline.points;
  const readings: ClockReading[] = [];

  for (let i = 0; i < points.length; i++) {
    const point = points[i];

    // Gravitational dilation at this position
    let gravitationalTickRate = 1.0;
    if (gravitationalBodies && gravitationalBodies.length > 0) {
      gravitationalTickRate = combinedDilation(
        gravitationalBodies,
        { x: point.x, y: point.y, z: point.z }
      );
    }

    // Kinematic dilation from velocity
    let velocityFractionC = 0;
    if (i > 0) {
      const prev = points[i - 1];
      const dt = point.pLo - prev.pLo;
      if (dt > 0) {
        const dx = point.x - prev.x;
        const dy = point.y - prev.y;
        const dz = point.z - prev.z;
        const distLy = Math.sqrt(dx * dx + dy * dy + dz * dz);
        // velocity in ly/pLo, convert to fraction of c
        // c = LIGHT_PICOLORENZO_LY ly/pLo, so v/c = (distLy/dt) / LIGHT_PICOLORENZO_LY
        const LIGHT_PICOLORENZO_LY = 0.008584070444; // c × π days in ly
        velocityFractionC = Math.min(
          0.9999,
          distLy / (dt * LIGHT_PICOLORENZO_LY)
        );
      }
    }

    const beta2 = velocityFractionC * velocityFractionC;
    const kinematicTickRate = beta2 < 1e-12 ? 1.0 : Math.sqrt(1 - beta2);

    // Combined tick rate
    const tickRate = gravitationalTickRate * kinematicTickRate;

    // Solar phase
    let solarPhaseVal: SolarPhase | null = null;
    let solarPhaseCont: number | null = null;
    let twilightZoneVal: TwilightZone | null = null;
    let solarAltDeg: number | null = null;

    if (observer) {
      const epochMs = point.pLo * PICOLORENZO_MS;
      const pos = solarPosition(observer, epochMs);
      solarPhaseVal = pos.zone === 'day' ? 1 : 0;
      solarPhaseCont = Math.max(0, Math.min(1, (pos.altitudeDeg + 90) / 180));
      twilightZoneVal = pos.zone;
      solarAltDeg = pos.altitudeDeg;
    }

    const driftPLo = point.properTimePLo - (point.pLo - points[0].pLo);

    readings.push({
      coordinateTimePLo: point.pLo,
      properTimePLo: point.properTimePLo,
      tickRate,
      x: point.x,
      y: point.y,
      z: point.z,
      velocityFractionC,
      gravitationalTickRate,
      kinematicTickRate,
      solarPhase: solarPhaseVal,
      solarPhaseContinuous: solarPhaseCont,
      twilightZone: twilightZoneVal,
      solarAltitudeDeg: solarAltDeg,
      gnosticTime: formatGnosticTime(point.properTimePLo),
      driftPLo,
    });
  }

  const totalProper = worldline.totalProperTimePLo;
  const totalCoord = worldline.totalCoordinateTimePLo;
  const averageTickRate = totalCoord > 0 ? totalProper / totalCoord : 1;
  const totalDrift = totalProper - totalCoord;

  let halfDay: SolarHalfDay | null = null;
  if (observer && points.length > 0) {
    halfDay = solarHalfDay(points[0].pLo, observer);
  }

  return {
    source: 'GnosisClock',
    id,
    label,
    worldline,
    readings,
    totalProperTimePLo: totalProper,
    totalCoordinateTimePLo: totalCoord,
    averageTickRate,
    totalDriftPLo: totalDrift,
    solarHalfDay: halfDay,
  };
}

// ─── Convenience Constructors ───────────────────────────────────────

/**
 * A clock sitting still on Earth's surface.
 * Ticks at ~1.0 (tiny gravitational dilation from Sun).
 * Has solar phase from observer location.
 */
export function earthSurfaceClock(
  id: string,
  label: string,
  observer: SolarObserver,
  startPLo: number,
  durationPLo: number,
  steps?: number
): Clock {
  const endPLo = startPLo + durationPLo;
  const wl = stationaryWorldline(id, label, { x: 0, y: 0, z: 0 }, startPLo, endPLo, steps ?? 55);
  return createClock({
    id,
    label,
    worldline: wl,
    gravitationalBodies: [{
      id: 'sun',
      label: 'Sun',
      massSolar: 1,
      positionLy: { x: 0, y: 0, z: 0 }, // heliocentric, Earth is at ~1 AU
    }],
    observer,
  });
}

/**
 * A clock orbiting a star (e.g., a planet's clock).
 * Experiences both kinematic and gravitational dilation.
 * Has solar phase from the orbit.
 */
export function orbitingClock(
  id: string,
  label: string,
  radiusLy: number,
  periodPLo: number,
  starMassSolar: number,
  startPLo: number,
  durationPLo: number,
  observer?: SolarObserver,
  steps?: number
): Clock {
  const endPLo = startPLo + durationPLo;
  const wl = circularOrbitWorldline(id, label, 0, 0, radiusLy, periodPLo, startPLo, endPLo, steps ?? 55);
  return createClock({
    id,
    label,
    worldline: wl,
    gravitationalBodies: [{
      id: `${id}-star`,
      label: `${label} star`,
      massSolar: starMassSolar,
      positionLy: { x: 0, y: 0, z: 0 },
    }],
    observer,
  });
}

/**
 * A clock falling freely (no forces, constant velocity).
 * Experiences kinematic dilation only.
 */
export function freeFallingClock(
  id: string,
  label: string,
  startPosition: { x: number; y: number; z: number },
  velocityLyPerPLo: { vx: number; vy: number; vz: number },
  startPLo: number,
  durationPLo: number,
  steps?: number
): Clock {
  const n = steps ?? 55;
  const stepPLo = durationPLo / Math.max(1, n - 1);
  const positions: { pLo: number; x: number; y: number; z: number }[] = [];

  for (let i = 0; i < n; i++) {
    const pLo = startPLo + i * stepPLo;
    const t = i * stepPLo;
    positions.push({
      pLo,
      x: startPosition.x + velocityLyPerPLo.vx * t,
      y: startPosition.y + velocityLyPerPLo.vy * t,
      z: startPosition.z + velocityLyPerPLo.vz * t,
    });
  }

  const wl = worldlineFromPositions(id, label, positions);
  return createClock({ id, label, worldline: wl });
}

// ─── Clock Comparison ───────────────────────────────────────────────

/**
 * Compare two clocks that share the same coordinate time window.
 *
 * Produces sync points where both clocks' readings are sampled at
 * the same coordinate time, showing how they drift relative to each other.
 */
export function compareClocks(clockA: Clock, clockB: Clock): ClockComparison {
  // Find overlapping coordinate time window
  const startA = clockA.readings[0]?.coordinateTimePLo ?? 0;
  const startB = clockB.readings[0]?.coordinateTimePLo ?? 0;
  const endA = clockA.readings[clockA.readings.length - 1]?.coordinateTimePLo ?? 0;
  const endB = clockB.readings[clockB.readings.length - 1]?.coordinateTimePLo ?? 0;

  const overlapStart = Math.max(startA, startB);
  const overlapEnd = Math.min(endA, endB);

  // Sample sync points from clock A's reading times that fall in overlap
  const syncPoints: ClockSyncPoint[] = [];

  for (const readingA of clockA.readings) {
    const t = readingA.coordinateTimePLo;
    if (t < overlapStart || t > overlapEnd) continue;

    // Interpolate clock B's proper time at this coordinate time
    const readingB = interpolateReading(clockB.readings, t);
    if (!readingB) continue;

    syncPoints.push({
      coordinateTimePLo: t,
      clockAProperTimePLo: readingA.properTimePLo,
      clockBProperTimePLo: readingB.properTimePLo,
      driftPLo: readingA.properTimePLo - readingB.properTimePLo,
      clockATickRate: readingA.tickRate,
      clockBTickRate: readingB.tickRate,
    });
  }

  const diffPLo = clockA.totalProperTimePLo - clockB.totalProperTimePLo;
  const diffYears = diffPLo * Math.PI / 365.25; // 1 pLo = π days

  return {
    source: 'GnosisClockComparison',
    clockA,
    clockB,
    aheadClockId: diffPLo >= 0 ? clockA.id : clockB.id,
    properTimeDifferencePLo: Math.abs(diffPLo),
    properTimeDifferenceYears: Math.abs(diffYears),
    properTimeRatio: clockB.totalProperTimePLo > 0
      ? clockA.totalProperTimePLo / clockB.totalProperTimePLo
      : 1,
    syncPoints,
  };
}

/**
 * Interpolate a clock reading at a given coordinate time.
 */
function interpolateReading(
  readings: readonly ClockReading[],
  targetPLo: number
): ClockReading | null {
  if (readings.length === 0) return null;
  if (readings.length === 1) return readings[0];

  // Find bracketing readings
  for (let i = 0; i < readings.length - 1; i++) {
    const a = readings[i];
    const b = readings[i + 1];
    if (targetPLo >= a.coordinateTimePLo && targetPLo <= b.coordinateTimePLo) {
      const range = b.coordinateTimePLo - a.coordinateTimePLo;
      const t = range > 0 ? (targetPLo - a.coordinateTimePLo) / range : 0;

      return {
        coordinateTimePLo: targetPLo,
        properTimePLo: a.properTimePLo + (b.properTimePLo - a.properTimePLo) * t,
        tickRate: a.tickRate + (b.tickRate - a.tickRate) * t,
        x: a.x + (b.x - a.x) * t,
        y: a.y + (b.y - a.y) * t,
        z: a.z + (b.z - a.z) * t,
        velocityFractionC: a.velocityFractionC + (b.velocityFractionC - a.velocityFractionC) * t,
        gravitationalTickRate: a.gravitationalTickRate + (b.gravitationalTickRate - a.gravitationalTickRate) * t,
        kinematicTickRate: a.kinematicTickRate + (b.kinematicTickRate - a.kinematicTickRate) * t,
        solarPhase: a.solarPhase,
        solarPhaseContinuous: a.solarPhaseContinuous !== null && b.solarPhaseContinuous !== null
          ? a.solarPhaseContinuous + (b.solarPhaseContinuous - a.solarPhaseContinuous) * t
          : a.solarPhaseContinuous,
        twilightZone: a.twilightZone,
        solarAltitudeDeg: a.solarAltitudeDeg !== null && b.solarAltitudeDeg !== null
          ? a.solarAltitudeDeg + (b.solarAltitudeDeg - a.solarAltitudeDeg) * t
          : a.solarAltitudeDeg,
        gnosticTime: formatGnosticTime(
          a.properTimePLo + (b.properTimePLo - a.properTimePLo) * t
        ),
        driftPLo: a.driftPLo + (b.driftPLo - a.driftPLo) * t,
      };
    }
  }

  // Outside range -- return nearest endpoint
  if (targetPLo <= readings[0].coordinateTimePLo) return readings[0];
  return readings[readings.length - 1];
}

// ─── Clock Queries ──────────────────────────────────────────────────

/**
 * What does this clock read right now?
 * Returns the most recent reading, or interpolates to a specific coordinate time.
 */
export function clockReadingAt(clock: Clock, coordinateTimePLo: number): ClockReading | null {
  return interpolateReading(clock.readings, coordinateTimePLo);
}

/**
 * Is this clock currently in daylight?
 */
export function clockIsDaylight(clock: Clock, coordinateTimePLo: number): boolean | null {
  const reading = clockReadingAt(clock, coordinateTimePLo);
  if (!reading || reading.solarPhase === null) return null;
  return reading.solarPhase === 1;
}

/**
 * How many ticks has this clock accumulated between two coordinate times?
 */
export function clockElapsed(
  clock: Clock,
  fromCoordinatePLo: number,
  toCoordinatePLo: number
): { readonly properTimePLo: number; readonly coordinateTimePLo: number; readonly averageTickRate: number } | null {
  const readingA = clockReadingAt(clock, fromCoordinatePLo);
  const readingB = clockReadingAt(clock, toCoordinatePLo);
  if (!readingA || !readingB) return null;

  const properDelta = readingB.properTimePLo - readingA.properTimePLo;
  const coordDelta = toCoordinatePLo - fromCoordinatePLo;

  return {
    properTimePLo: properDelta,
    coordinateTimePLo: coordDelta,
    averageTickRate: coordDelta > 0 ? properDelta / coordDelta : 1,
  };
}

// ─── Clock Mesh ─────────────────────────────────────────────────────

/**
 * A cell in a clock mesh: one clock's state at one point on the grid.
 */
export interface ClockMeshCell {
  readonly lat: number;
  readonly lon: number;
  readonly epochMs: number;
  /** Solar tick rate: continuous 0-1 solar phase as a "temporal temperature." */
  readonly solarTickRate: number;
  /** Binary phase: 1 = lit (heating), 0 = dark (cooling). */
  readonly solarPhase: SolarPhase;
  /** Twilight zone for this cell. */
  readonly twilightZone: TwilightZone;
  /** Solar altitude in degrees. */
  readonly solarAltitudeDeg: number;
  /** How fast time "feels" here relative to the grid mean. */
  readonly relativeTickRate: number;
}

/**
 * A mesh of clocks across a geographic region at a given time.
 *
 * Each cell is a clock. The mesh shows how "time flows" across space:
 * lit cells are in their fast phase (solar heating, active weather),
 * dark cells are in their slow phase (radiative cooling, convective
 * shutdown). The terminator sweeping across the mesh is time's edge.
 */
export interface ClockMesh {
  readonly source: 'GnosisClockMesh';
  readonly epochMs: number;
  readonly resolutionDeg: number;
  readonly latMin: number;
  readonly latMax: number;
  readonly lonMin: number;
  readonly lonMax: number;
  readonly latCount: number;
  readonly lonCount: number;
  readonly cells: readonly ClockMeshCell[];
  /** Fraction of cells in daylight. */
  readonly litFraction: number;
  /** Mean solar tick rate across the mesh. */
  readonly meanSolarTickRate: number;
  /** The "temporal gradient" -- how much tick rate varies across the mesh. */
  readonly tickRateRange: number;
}

export interface CreateClockMeshInput {
  readonly epochMs?: number;
  readonly resolutionDeg?: number;
  readonly latMin?: number;
  readonly latMax?: number;
  readonly lonMin?: number;
  readonly lonMax?: number;
}

/**
 * Create a mesh of clocks across a geographic region.
 *
 * Each cell computes the solar position and maps it to a "tick rate" --
 * a continuous value representing how fast time flows at that point.
 * The lit side ticks at solarPhaseContinuous > 0.5 (heating phase),
 * the dark side at < 0.5 (cooling phase). The terminator is the
 * 0.5 contour.
 *
 * For storms.watch: overlay this on the storm field to see how the
 * diurnal cycle sweeps across the system. The storm's convective
 * intensity follows this temporal rhythm.
 */
export function createClockMesh(input: CreateClockMeshInput = {}): ClockMesh {
  const {
    epochMs = Date.now(),
    resolutionDeg = 1,
    latMin = -90,
    latMax = 90,
    lonMin = -180,
    lonMax = 180,
  } = input;

  const latCount = Math.round((latMax - latMin) / resolutionDeg) + 1;
  const lonCount = Math.round((lonMax - lonMin) / resolutionDeg);

  const cells: ClockMeshCell[] = [];
  let litCount = 0;
  let totalTickRate = 0;

  for (let li = 0; li < latCount; li++) {
    const lat = latMin + li * resolutionDeg;
    for (let lo = 0; lo < lonCount; lo++) {
      const lon = lonMin + lo * resolutionDeg;

      const pos = solarPosition({ latitudeDeg: lat, longitudeDeg: lon }, epochMs);
      const phase: SolarPhase = pos.zone === 'day' ? 1 : 0;
      // Continuous solar tick rate: maps altitude [-90, 90] to [0, 1]
      const solarTickRate = Math.max(0, Math.min(1, (pos.altitudeDeg + 90) / 180));

      if (phase === 1) litCount++;
      totalTickRate += solarTickRate;

      cells.push({
        lat,
        lon,
        epochMs,
        solarTickRate,
        solarPhase: phase,
        twilightZone: pos.zone,
        solarAltitudeDeg: pos.altitudeDeg,
        relativeTickRate: 0, // filled in after mean is computed
      });
    }
  }

  const totalCells = cells.length;
  const meanRate = totalCells > 0 ? totalTickRate / totalCells : 0.5;

  // Fill in relative tick rates
  let minRate = Infinity;
  let maxRate = -Infinity;
  const finalCells = cells.map(cell => {
    const relative = meanRate > 0 ? cell.solarTickRate / meanRate : 1;
    if (cell.solarTickRate < minRate) minRate = cell.solarTickRate;
    if (cell.solarTickRate > maxRate) maxRate = cell.solarTickRate;
    return { ...cell, relativeTickRate: relative };
  });

  return {
    source: 'GnosisClockMesh',
    epochMs,
    resolutionDeg,
    latMin,
    latMax,
    lonMin,
    lonMax,
    latCount,
    lonCount,
    cells: finalCells,
    litFraction: totalCells > 0 ? litCount / totalCells : 0,
    meanSolarTickRate: meanRate,
    tickRateRange: maxRate - minRate,
  };
}

/**
 * Create a clock mesh centered on a storm's position.
 * Default: 20-degree box around the storm center at 0.5-degree resolution.
 */
export function stormClockMesh(
  stormLat: number,
  stormLon: number,
  epochMs?: number,
  radiusDeg?: number,
  resolutionDeg?: number
): ClockMesh {
  const r = radiusDeg ?? 10;
  const res = resolutionDeg ?? 0.5;
  return createClockMesh({
    epochMs,
    resolutionDeg: res,
    latMin: Math.max(-90, stormLat - r),
    latMax: Math.min(90, stormLat + r),
    lonMin: stormLon - r,
    lonMax: stormLon + r,
  });
}

/**
 * At what coordinate time does this clock's proper time reading reach a given value?
 * Inverse lookup: "when does my clock say X?"
 */
export function clockReachesProperTime(
  clock: Clock,
  targetProperTimePLo: number
): number | null {
  const readings = clock.readings;
  if (readings.length === 0) return null;

  for (let i = 0; i < readings.length - 1; i++) {
    const a = readings[i];
    const b = readings[i + 1];
    if (targetProperTimePLo >= a.properTimePLo && targetProperTimePLo <= b.properTimePLo) {
      const range = b.properTimePLo - a.properTimePLo;
      const t = range > 0 ? (targetProperTimePLo - a.properTimePLo) / range : 0;
      return a.coordinateTimePLo + (b.coordinateTimePLo - a.coordinateTimePLo) * t;
    }
  }

  return null; // target proper time not within this clock's range
}
