// @a0n/gnosis/src/clock-topology-coordinator.ts
/**
 * Clock Topology Coordinator -- pure functions that convert observations
 * into distributed clock meshes and compute topology deltas.
 *
 * Takes raw ClockObservation records (from AeonPID heartbeats) and
 * builds the temporal topology of the network. Compares successive
 * meshes to detect structural changes (nodes joining/leaving, edges
 * promoting from VENT to PROCESS or demoting).
 *
 * @packageDocumentation
 */

import type {
  ProcessClock,
  ProcessPhase,
  DistributedClockMesh,
} from './clock-distributed';
import { createDistributedClockMesh } from './clock-distributed';
import { coloToLatLon } from './colo-coordinates';
import { toPicolorenzos, PICOLORENZO_MS } from './math';

// ─── Observation ────────────────────────────────────────────────────

/** A single heartbeat observation from an AeonPID. */
export interface ClockObservation {
  readonly pidId: string;
  readonly label: string;
  readonly colo: string;
  readonly clockPLo: number;
  readonly rttMs: number;
  readonly phase: ProcessPhase;
  readonly requestRate: number;
}

// ─── Topology Delta ─────────────────────────────────────────────────

/** Structural difference between two successive clock meshes. */
export interface TopologyDelta {
  readonly addedNodes: readonly string[];
  readonly removedNodes: readonly string[];
  readonly promotedEdges: readonly { fromId: string; toId: string }[];
  readonly demotedEdges: readonly { fromId: string; toId: string }[];
  readonly connectivityChange: number;
}

// ─── Conversion ─────────────────────────────────────────────────────

/**
 * Convert a single observation into a ProcessClock.
 * Returns null if the colo code has no known coordinates.
 */
export function observationToProcessClock(obs: ClockObservation): ProcessClock | null {
  const coords = coloToLatLon(obs.colo);
  if (!coords) return null;

  return {
    id: obs.pidId,
    label: obs.label,
    location: {
      lat: coords.lat,
      lon: coords.lon,
      region: obs.colo.toUpperCase(),
    },
    phase: obs.phase,
    localEpochMs: obs.clockPLo * PICOLORENZO_MS,
    lastHeartbeatMs: Date.now(),
    requestRate: obs.requestRate,
  };
}

// ─── Build Topology ─────────────────────────────────────────────────

/**
 * Build a DistributedClockMesh from raw observations.
 *
 * Unknown colo codes are silently dropped. The optional toleranceMs
 * controls the PROCESS/VENT edge classification threshold.
 */
export function buildClockTopology(
  observations: readonly ClockObservation[],
  toleranceMs?: number,
): DistributedClockMesh {
  const clocks: ProcessClock[] = [];
  for (const obs of observations) {
    const clock = observationToProcessClock(obs);
    if (clock) clocks.push(clock);
  }
  return createDistributedClockMesh(clocks, toleranceMs);
}

// ─── Topology Delta ─────────────────────────────────────────────────

/**
 * Compute the structural delta between two successive meshes.
 *
 * - addedNodes: ids present in `next` but not `prev`
 * - removedNodes: ids present in `prev` but not `next`
 * - promotedEdges: edges that were VENT in `prev` but PROCESS in `next`
 * - demotedEdges: edges that were PROCESS in `prev` but VENT in `next`
 * - connectivityChange: delta in physicalOrderingFraction
 */
export function topologyDelta(
  prev: DistributedClockMesh,
  next: DistributedClockMesh,
): TopologyDelta {
  const prevIds = new Set(prev.clocks.map(c => c.id));
  const nextIds = new Set(next.clocks.map(c => c.id));

  const addedNodes = next.clocks
    .filter(c => !prevIds.has(c.id))
    .map(c => c.id);
  const removedNodes = prev.clocks
    .filter(c => !nextIds.has(c.id))
    .map(c => c.id);

  // Build edge type maps keyed by "fromId:toId" (sorted pair)
  function edgeKey(fromId: string, toId: string): string {
    return fromId < toId ? `${fromId}:${toId}` : `${toId}:${fromId}`;
  }

  const prevEdgeTypes = new Map<string, 'PROCESS' | 'VENT'>();
  for (const e of prev.edges) {
    prevEdgeTypes.set(edgeKey(e.fromId, e.toId), e.edgeType);
  }

  const nextEdgeTypes = new Map<string, 'PROCESS' | 'VENT'>();
  for (const e of next.edges) {
    nextEdgeTypes.set(edgeKey(e.fromId, e.toId), e.edgeType);
  }

  const promotedEdges: { fromId: string; toId: string }[] = [];
  const demotedEdges: { fromId: string; toId: string }[] = [];

  for (const [key, nextType] of nextEdgeTypes) {
    const prevType = prevEdgeTypes.get(key);
    if (prevType === 'VENT' && nextType === 'PROCESS') {
      const [fromId, toId] = key.split(':') as [string, string];
      promotedEdges.push({ fromId, toId });
    } else if (prevType === 'PROCESS' && nextType === 'VENT') {
      const [fromId, toId] = key.split(':') as [string, string];
      demotedEdges.push({ fromId, toId });
    }
  }

  return {
    addedNodes,
    removedNodes,
    promotedEdges,
    demotedEdges,
    connectivityChange: next.physicalOrderingFraction - prev.physicalOrderingFraction,
  };
}
