// @a0n/gnosis/src/colo-coordinates.ts
/**
 * Cloudflare datacenter coordinates by IATA code.
 *
 * Static map of the top ~50 major Cloudflare edge locations to their
 * geographic coordinates. Used by clock topology modules to convert
 * colo codes into lat/lon for propagation delay computation.
 *
 * @packageDocumentation
 */

/** Cloudflare datacenter coordinates by IATA code. */
export const COLO_COORDINATES: Readonly<Record<string, { readonly lat: number; readonly lon: number; readonly city: string }>> = {
  SFO: { lat: 37.62, lon: -122.38, city: 'San Francisco' },
  LAX: { lat: 33.94, lon: -118.41, city: 'Los Angeles' },
  SEA: { lat: 47.45, lon: -122.31, city: 'Seattle' },
  ORD: { lat: 41.98, lon: -87.90, city: 'Chicago' },
  IAD: { lat: 38.94, lon: -77.46, city: 'Washington DC' },
  EWR: { lat: 40.69, lon: -74.17, city: 'Newark' },
  ATL: { lat: 33.64, lon: -84.43, city: 'Atlanta' },
  DFW: { lat: 32.90, lon: -97.04, city: 'Dallas' },
  MIA: { lat: 25.79, lon: -80.29, city: 'Miami' },
  YYZ: { lat: 43.68, lon: -79.63, city: 'Toronto' },
  LHR: { lat: 51.47, lon: -0.46, city: 'London' },
  CDG: { lat: 49.01, lon: 2.55, city: 'Paris' },
  AMS: { lat: 52.31, lon: 4.76, city: 'Amsterdam' },
  FRA: { lat: 50.03, lon: 8.57, city: 'Frankfurt' },
  ARN: { lat: 59.65, lon: 17.94, city: 'Stockholm' },
  MAD: { lat: 40.47, lon: -3.56, city: 'Madrid' },
  MXP: { lat: 45.63, lon: 8.72, city: 'Milan' },
  WAW: { lat: 52.17, lon: 20.97, city: 'Warsaw' },
  NRT: { lat: 35.76, lon: 140.39, city: 'Tokyo' },
  HKG: { lat: 22.31, lon: 113.91, city: 'Hong Kong' },
  SIN: { lat: 1.36, lon: 103.99, city: 'Singapore' },
  ICN: { lat: 37.46, lon: 126.44, city: 'Seoul' },
  BOM: { lat: 19.09, lon: 72.87, city: 'Mumbai' },
  SYD: { lat: -33.95, lon: 151.18, city: 'Sydney' },
  MEL: { lat: -37.67, lon: 144.84, city: 'Melbourne' },
  GRU: { lat: -23.43, lon: -46.47, city: 'São Paulo' },
  JNB: { lat: -26.13, lon: 28.24, city: 'Johannesburg' },
  DXB: { lat: 25.25, lon: 55.36, city: 'Dubai' },
  CPT: { lat: -33.97, lon: 18.60, city: 'Cape Town' },
  KIX: { lat: 34.43, lon: 135.23, city: 'Osaka' },
  TPE: { lat: 25.08, lon: 121.23, city: 'Taipei' },
  DEL: { lat: 28.57, lon: 77.09, city: 'Delhi' },
  CGK: { lat: -6.13, lon: 106.66, city: 'Jakarta' },
  BKK: { lat: 13.69, lon: 100.75, city: 'Bangkok' },
  MNL: { lat: 14.51, lon: 121.02, city: 'Manila' },
  SCL: { lat: -33.39, lon: -70.79, city: 'Santiago' },
  BOG: { lat: 4.70, lon: -74.15, city: 'Bogotá' },
  LIS: { lat: 38.77, lon: -9.13, city: 'Lisbon' },
  HEL: { lat: 60.32, lon: 24.96, city: 'Helsinki' },
  OSL: { lat: 60.19, lon: 11.10, city: 'Oslo' },
  ZRH: { lat: 47.46, lon: 8.55, city: 'Zurich' },
  VIE: { lat: 48.11, lon: 16.57, city: 'Vienna' },
  PRG: { lat: 50.10, lon: 14.26, city: 'Prague' },
  IST: { lat: 41.26, lon: 28.74, city: 'Istanbul' },
  TLV: { lat: 32.01, lon: 34.89, city: 'Tel Aviv' },
  DOH: { lat: 25.26, lon: 51.57, city: 'Doha' },
  YVR: { lat: 49.19, lon: -123.18, city: 'Vancouver' },
  PDX: { lat: 45.59, lon: -122.60, city: 'Portland' },
  DEN: { lat: 39.86, lon: -104.67, city: 'Denver' },
  PHX: { lat: 33.43, lon: -112.01, city: 'Phoenix' },
};

/** Look up coordinates for a CF datacenter code. Returns null for unknown colos. */
export function coloToLatLon(colo: string): { lat: number; lon: number; city: string } | null {
  return COLO_COORDINATES[colo.toUpperCase()] ?? null;
}

/** All known datacenter codes. */
export const KNOWN_COLOS = Object.keys(COLO_COORDINATES);
