// @a0n/gnosis/src/combat-topology.ts
/**
 * Combat Topology & Fighting Techniques
 *
 * Provides metric definitions for zero-sum positional bounding. Models
 * advantage and leverage as local network isolation bounds and structural
 * geometric dominance in a two-agent system.
 */

export interface CombatPosition {
  /** Degrees of mechanical freedom available (available paths to safely route energy) */
  degreesOfFreedom: number;
  /** The energy currently trapped in the agent's local geometry (cringe / damage proxy) */
  trappedHeat: number;
  /** Leverage: Structural advantage. Distance from the opponent's geometric center of gravity. */
  leverageFactor: number;
}

export type CombatAdvantage = 'dominant' | 'neutral' | 'submissive' | 'checkmate';

/**
 * Determines the topological advantage state of Agent A relative to Agent B.
 * 
 * In Combat Topology, advantage is purely the delta in accessible degrees of freedom,
 * multiplied by the leverage factor (how deep one agent penetrates the other's cover space).
 */
export function calculateCombatAdvantage(
  agentA: CombatPosition,
  agentB: CombatPosition
): CombatAdvantage {
  try {
    const freedomDelta = agentA.degreesOfFreedom - agentB.degreesOfFreedom;
    const leverageDelta = agentA.leverageFactor - agentB.leverageFactor;

    // Checkmate: Opponent has zero degrees of freedom (cannot VENT or FOLD).
    if (agentB.degreesOfFreedom === 0 && agentA.degreesOfFreedom > 0) {
      return 'checkmate';
    }

    if (freedomDelta > 0 && leverageDelta > 0) {
      return 'dominant';
    } else if (freedomDelta < 0 && leverageDelta < 0) {
      return 'submissive';
    }

    return 'neutral';

  } catch (error) {
    throw new Error(`[calculateCombatAdvantage] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * The kinetic transfer of thermodynamic force through a topological limb.
 * A strike reduces the target's degrees of freedom or increases their trapped heat.
 */
export function executeTopologicalStrike(
  attacker: CombatPosition,
  target: CombatPosition,
  kineticForce: number
): CombatPosition {
  // Leverage amplifies the injection of trapped heat
  const appliedHeat = kineticForce * Math.max(1, attacker.leverageFactor);
  
  // High heat forces the target to collapse degrees of freedom to maintain structural integrity
  const freedomLoss = Math.floor(appliedHeat / 10);
  
  return {
    ...target,
    trappedHeat: target.trappedHeat + appliedHeat,
    degreesOfFreedom: Math.max(0, target.degreesOfFreedom - freedomLoss)
  };
}
