// @a0n/gnosis/src/compressor-cascade.ts
/**
 * compressor-cascade.ts — gnosis-local mirror of the jet-engine **compressor cascade**.
 *
 * Byte-identical API to `open-source/aether/src/wasm-simd/compressor-cascade.ts` and the Rust
 * `gnosis-engine-core` crate, so every runtime carries the same proof-backed law. A runtime
 * pipeline is a list of {@link Stage}s, each with a *measured* speedup / compression ratio. The
 * jet-engine compressor law says the pipeline's overall ratio is the **product** of its stage
 * ratios ({@link overallRatio}), and stacking two pipelines **multiplies** their ratios — the
 * `(List, ++) -> (·, ×)` monoid homomorphism ({@link compose} + {@link homomorphismHolds}).
 *
 * Proof backing (gnosis-math, axiom-clean):
 *   - Gnosis.MathJetEngine.overallRatio                — overall ratio = product of stages
 *   - Gnosis.MathJetEngine.overallRatio_append         — stacking multiplies (Nat)
 *   - Gnosis.CompressorCascadeRuntime.prodOver_append  — stacking multiplies (any monoid; covers the
 *                                                        measured rational ratios used here)
 *   - Gnosis.MathJetEngine.jet_is_steady_ripcord       — the cascade product is the ripcord spin
 *
 * Scope: the compressor cascade only — not the spool governor or additive turbofan thrust.
 */

/** Canonical Lean theorem IDs backing this cascade — kept byte-identical to the Rust crate. */
export const MATH_JET_ENGINE_MASTER_THEOREM_ID = 'Gnosis.MathJetEngine.math_jet_engine_master';
export const OVERALL_RATIO_APPEND_THEOREM_ID = 'Gnosis.MathJetEngine.overallRatio_append';
export const JET_IS_STEADY_RIPCORD_THEOREM_ID = 'Gnosis.MathJetEngine.jet_is_steady_ripcord';
export const PRODOVER_APPEND_THEOREM_ID = 'Gnosis.CompressorCascadeRuntime.prodOver_append';
export const RUNTIME_CASCADE_THEOREM_ID =
  'Gnosis.CompressorCascadeRuntime.runtime_compressor_cascade_composes';

/** The theorem-ID set every cascade report carries. */
export const CASCADE_THEOREM_IDS = Object.freeze([
  MATH_JET_ENGINE_MASTER_THEOREM_ID,
  OVERALL_RATIO_APPEND_THEOREM_ID,
  JET_IS_STEADY_RIPCORD_THEOREM_ID,
  PRODOVER_APPEND_THEOREM_ID,
  RUNTIME_CASCADE_THEOREM_ID,
] as const);

/** The E₈ Coxeter number — the overall ratio of the canonical `[2,3,5]` compressor. */
export const E8_COXETER_NUMBER = 30;

/**
 * A compressor stage: a named pipeline transform with a measured ratio.
 * `ratio` is the stage's speedup / compression factor in isolation (`> 1` accelerates, `1` is a
 * clean pass-through — the identity of the cascade).
 */
export interface Stage {
  readonly name: string;
  readonly ratio: number;
}

/**
 * Handles the gnosis stage workflow.
 */
export function stage(name: string, ratio: number): Stage {
  return { name, ratio };
}

/**
 * Overall pressure ratio of a stack = **product** of the stage ratios.
 * Mirrors `Gnosis.MathJetEngine.overallRatio` (`prodOver (·*·) 1`). Empty cascade = `1` (identity).
 */
export function overallRatio(stages: readonly Stage[]): number {
  return stages.reduce((acc, s) => acc * s.ratio, 1);
}

/**
 * **Append homomorphism**: stacking two compressors concatenates their stage lists, so that
 * `overallRatio(compose(a, b)) === overallRatio(a) * overallRatio(b)`.
 * Mirrors `Gnosis.MathJetEngine.overallRatio_append`.
 */
export function compose(a: readonly Stage[], b: readonly Stage[]): Stage[] {
  return [...a, ...b];
}

/**
 * The append homomorphism holds within relative tolerance `tol`:
 * `|overall(compose(a,b)) - overall(a)·overall(b)| <= tol · max(overall(compose(a,b)), 1)`.
 */
export function homomorphismHolds(a: readonly Stage[], b: readonly Stage[], tol = 1e-9): boolean {
  const lhs = overallRatio(compose(a, b));
  const rhs = overallRatio(a) * overallRatio(b);
  return Math.abs(lhs - rhs) <= tol * Math.max(Math.abs(lhs), 1);
}

/**
 * A measured cascade report — the honest per-stage-vs-composed record a bench fills in.
 * `predictedProduct` is the design-law prediction (product of the **isolated** stage ratios);
 * `measuredEndToEnd` is the **real** composed speedup. The residual `measured / predicted` is `1`
 * when the product law predicts the pipeline exactly.
 */
export interface CascadeReport {
  readonly label: string;
  readonly stages: readonly Stage[];
  readonly predictedProduct: number;
  readonly measuredEndToEnd: number;
  readonly theoremIds: readonly string[];
}

/**
 * Handles the gnosis cascade Report workflow.
 */
export function cascadeReport(
  label: string,
  stages: readonly Stage[],
  measuredEndToEnd: number,
): CascadeReport {
  return {
    label,
    stages,
    predictedProduct: overallRatio(stages),
    measuredEndToEnd,
    theoremIds: CASCADE_THEOREM_IDS,
  };
}

/** Residual = measured / predicted (`1` ⇒ the product law predicts end-to-end exactly). */
export function residual(r: CascadeReport): number {
  return r.predictedProduct === 0 ? 0 : r.measuredEndToEnd / r.predictedProduct;
}

/** The product law predicts the measured end-to-end speedup within relative tolerance `tol`. */
export function lawHolds(r: CascadeReport, tol = 1e-9): boolean {
  return (
    Math.abs(r.measuredEndToEnd - r.predictedProduct) <=
    tol * Math.max(Math.abs(r.predictedProduct), 1)
  );
}

/** A one-line, emoji-free summary for bench output. */
export function summarize(r: CascadeReport): string {
  const stages = r.stages.map((s) => `${s.name}=${s.ratio.toFixed(3)}x`).join(' * ');
  return (
    `${r.label}: stages[${stages}] ` +
    `predicted=${r.predictedProduct.toFixed(4)}x measured=${r.measuredEndToEnd.toFixed(4)}x ` +
    `residual=${residual(r).toFixed(3)}`
  );
}
