// @a0n/gnosis/src/continuous-harris-handlers.ts
/**
 * Continuous Harris GG Topology Handlers
 *
 * Handlers for Harris certificate synthesis, observable inference,
 * small set generation, Lyapunov synthesis, and Lean artifact generation.
 */

import type { GnosisRegistry } from './runtime/registry';
import {
  synthesizeSmallSet,
  inferObservable,
  synthesizeLyapunov,
  synthesizeMinorization,
  synthesizeHarrisCertificate,
  inferKernelFamily,
  generateHarrisLean,
  type HarrisCertificate,
  type ObservableKind,
} from './continuous-harris';

/**
 * Handles the gnosis register Continuous Harris Handlers workflow.
 */
export function registerContinuousHarrisHandlers(
  registry: GnosisRegistry
): void {
  // Infer observable from node properties
  registry.register('ObservableInference', async (_payload, props) => {
    return inferObservable(props);
  });

  // Synthesize small set from node properties
  registry.register('SmallSetSynthesizer', async (_payload, props) => {
    const nodeId = props.nodeId ?? 'state';
    const observableKind = (props.observable_kind ??
      'queue-length') as ObservableKind;
    return synthesizeSmallSet(nodeId, observableKind, props);
  });

  // Synthesize Lyapunov function
  registry.register('LyapunovSynthesizer', async (_payload, props) => {
    const observable = inferObservable(props);
    return synthesizeLyapunov(observable, props);
  });

  // Synthesize minorization data
  registry.register('MinorizationSynthesizer', async (_payload, props) => {
    const nodeId = props.nodeId ?? 'state';
    const observableKind = (props.observable_kind ??
      'queue-length') as ObservableKind;
    const smallSet = synthesizeSmallSet(nodeId, observableKind, props);
    return synthesizeMinorization(smallSet, props);
  });

  // Full Harris certificate synthesis
  registry.register('HarrisCertificateSynthesizer', async (_payload, props) => {
    const nodeId = props.nodeId ?? 'state';
    return synthesizeHarrisCertificate(nodeId, props);
  });

  // Infer kernel family
  registry.register('KernelFamilyInference', async (_payload, props) => {
    return inferKernelFamily(props);
  });

  // Generate Lean artifact from Harris certificate
  registry.register('HarrisLeanGenerator', async (payload) => {
    const cert = payload as HarrisCertificate;
    if (!cert?.nodeId) {
      return { error: 'No Harris certificate provided' };
    }
    const lean = generateHarrisLean(cert);
    return {
      lean,
      theoremName: cert.theoremName,
      moduleName: cert.moduleName,
      fullyVerified: cert.fullyVerified,
    };
  });

  // Full pipeline: properties → certificate → Lean
  registry.register('HarrisPipeline', async (_payload, props) => {
    const nodeId = props.nodeId ?? 'state';
    const cert = synthesizeHarrisCertificate(nodeId, props);
    const lean = generateHarrisLean(cert);
    return {
      certificate: cert,
      lean,
      status: cert.fullyVerified ? 'CERTIFIED' : 'PARTIAL',
    };
  });
}

export const HARRIS_KERNEL_FAMILIES = [
  'queue-support',
  'diffusion',
  'jump-diffusion',
  'piecewise-deterministic',
  'renewal',
  'general-feller',
] as const;
