// @a0n/gnosis/src/continuous-harris.ts
/**
 * continuous-harris.ts -- Compiler-driven measurable Harris packages
 *
 * Synthesize measurable small sets from topology syntax. Support richer
 * observable/Lyapunov families. Move beyond queue-support kernels toward
 * non-queue measurable kernels with explicit Harris/minorization data.
 *
 * Roadmap item #10: extends the existing stability synthesis pipeline.
 */

// ============================================================================
// Small set synthesis -- from topology to measurable sets
// ============================================================================

export type SmallSetKind =
  | 'queue-boundary'
  | 'level-set'
  | 'capacity-sublevel'
  | 'compact-sublevel'
  | 'finite-union'
  | 'synthesized';

export interface SmallSet {
  kind: SmallSetKind;
  /** Human-readable description */
  description: string;
  /** Parameter that defines the set boundary */
  boundaryParam: number;
  /** Lean expression for the set */
  leanExpr: string;
  /** Whether this small set has been verified */
  verified: boolean;
}

/**
 * Handles the gnosis synthesize Small Set workflow.
 */
export function synthesizeSmallSet(
  nodeId: string,
  observableKind: ObservableKind,
  properties: Record<string, string>
): SmallSet {
  try {
    // Queue boundary: C = {x : queue_length(x) <= B}
    if (
      observableKind === 'queue-length' ||
      observableKind === 'bounded-affine-queue'
    ) {
      const boundary = parseFloat(
        properties['queue_boundary'] ?? properties['capacity'] ?? '10'
      );
      return {
        kind: 'queue-boundary',
        description: `{x : queue_length(x) <= ${boundary}}`,
        boundaryParam: boundary,
        leanExpr: `{x | x.queueLength ≤ ${boundary}}`,
        verified: true,
      };
    }

    // Level set: C = {x : V(x) <= c}
    if (observableKind === 'lyapunov' || observableKind === 'energy') {
      const level = parseFloat(
        properties['level'] ?? properties['energy_threshold'] ?? '1'
      );
      return {
        kind: 'level-set',
        description: `{x : V(x) <= ${level}}`,
        boundaryParam: level,
        leanExpr: `{x | V x ≤ ${level}}`,
        verified: true,
      };
    }

    // Capacity sublevel: C = {x : cap(x) <= K}
    if (observableKind === 'capacity') {
      const cap = parseFloat(properties['capacity'] ?? '1');
      return {
        kind: 'capacity-sublevel',
        description: `{x : capacity(x) <= ${cap}}`,
        boundaryParam: cap,
        leanExpr: `{x | capacity x ≤ ${cap}}`,
        verified: true,
      };
    }

    // Compact sublevel for general observables
    if (observableKind === 'compact' || observableKind === 'bounded') {
      const bound = parseFloat(
        properties['bound'] ?? properties['diameter'] ?? '1'
      );
      return {
        kind: 'compact-sublevel',
        description: `{x : ||x|| <= ${bound}}`,
        boundaryParam: bound,
        leanExpr: `Metric.closedBall 0 ${bound}`,
        verified: true,
      };
    }

    // Generic fallback: synthesize from drift gap
    const driftGap = parseFloat(properties['drift_gap'] ?? '0.1');
    return {
      kind: 'synthesized',
      description: `synthesized small set (drift_gap=${driftGap})`,
      boundaryParam: 1 / driftGap,
      leanExpr: `{x | f x ≤ ${1 / driftGap}}`,
      verified: false,
    };

  } catch (error) {
    throw new Error(`[synthesizeSmallSet] ${error instanceof Error ? error.message : String(error)}`);
  }
}

// ============================================================================
// Observable families -- richer than bounded affine queue
// ============================================================================

export type ObservableKind =
  | 'queue-length'
  | 'bounded-affine-queue'
  | 'lyapunov'
  | 'energy'
  | 'capacity'
  | 'compact'
  | 'bounded'
  | 'polynomial'
  | 'log-barrier'
  | 'piecewise-linear'
  | 'custom';

export interface Observable {
  kind: ObservableKind;
  /** Human-readable expression */
  expression: string;
  /** Lean expression */
  leanExpr: string;
  /** Scale factor for normalized drift */
  scale: number;
  /** Offset for normalized drift */
  offset: number;
}

/**
 * Handles the gnosis infer Observable workflow.
 */
export function inferObservable(
  properties: Record<string, string>
): Observable {
  const kind = (properties['observable_kind'] ??
    'queue-length') as ObservableKind;
  const expression = properties['observable'] ?? 'queue_length';
  const scale = parseFloat(properties['observable_scale'] ?? '1');
  const offset = parseFloat(properties['observable_offset'] ?? '0');

  return {
    kind,
    expression,
    leanExpr: observableToLean(kind, expression),
    scale,
    offset,
  };
}

function observableToLean(kind: ObservableKind, expression: string): string {
  try {
    switch (kind) {
      case 'queue-length':
      case 'bounded-affine-queue':
        return `fun x => (x.queueLength : ℝ)`;
      case 'lyapunov':
        return `fun x => V x`;
      case 'energy':
        return `fun x => E x`;
      case 'capacity':
        return `fun x => capacity x`;
      case 'polynomial':
        return `fun x => ${expression}`;
      case 'log-barrier':
        return `fun x => -Real.log (1 - ${expression})`;
      case 'piecewise-linear':
        return `fun x => piecewiseLinear x`;
      default:
        return `fun x => ${expression}`;
    }

  } catch (error) {
    throw new Error(`[observableToLean] ${error instanceof Error ? error.message : String(error)}`);
  }
}

// ============================================================================
// Lyapunov families -- drift witness synthesis
// ============================================================================

export type LyapunovKind =
  | 'affine'
  | 'quadratic'
  | 'polynomial'
  | 'log-barrier'
  | 'piecewise-linear'
  | 'exponential'
  | 'custom';

export interface LyapunovFunction {
  kind: LyapunovKind;
  expression: string;
  leanExpr: string;
  /** Drift bound: E[V(X_{n+1}) | X_n = x] <= λV(x) + b for x ∉ C */
  lambda: number;
  /** Additive constant in drift condition */
  b: number;
}

/**
 * Handles the gnosis synthesize Lyapunov workflow.
 */
export function synthesizeLyapunov(
  observable: Observable,
  properties: Record<string, string>
): LyapunovFunction {
  try {
    const driftGap = parseFloat(properties['drift_gap'] ?? '0.1');
    const lambda = 1 - driftGap;

    switch (observable.kind) {
      case 'queue-length':
      case 'bounded-affine-queue': {
        const scale = observable.scale;
        const offset = observable.offset;
        return {
          kind: 'affine',
          expression: `${scale} * queue_length + ${offset}`,
          leanExpr: `fun x => ${scale} * x.queueLength + ${offset}`,
          lambda,
          b: parseFloat(properties['drift_b'] ?? '1'),
        };
      }

      case 'energy':
      case 'lyapunov': {
        return {
          kind: 'quadratic',
          expression: `V(x)^2`,
          leanExpr: `fun x => (V x) ^ 2`,
          lambda,
          b: parseFloat(properties['drift_b'] ?? '1'),
        };
      }

      case 'polynomial': {
        const degree = parseInt(properties['polynomial_degree'] ?? '2', 10);
        return {
          kind: 'polynomial',
          expression: `V(x)^${degree}`,
          leanExpr: `fun x => (${observable.expression}) ^ ${degree}`,
          lambda,
          b: parseFloat(properties['drift_b'] ?? '1'),
        };
      }

      case 'compact':
      case 'bounded': {
        return {
          kind: 'log-barrier',
          expression: `-log(1 - ||x||/R)`,
          leanExpr: `fun x => -Real.log (1 - ‖x‖ / R)`,
          lambda,
          b: parseFloat(properties['drift_b'] ?? '1'),
        };
      }

      default: {
        return {
          kind: 'affine',
          expression: `${observable.expression}`,
          leanExpr: observable.leanExpr,
          lambda,
          b: parseFloat(properties['drift_b'] ?? '1'),
        };
      }
    }

  } catch (error) {
    throw new Error(`[synthesizeLyapunov] ${error instanceof Error ? error.message : String(error)}`);
  }
}

// ============================================================================
// Minorization synthesis -- kernel lower bounds
// ============================================================================

export type MinorizationKind =
  | 'dirac'
  | 'uniform-on-small-set'
  | 'lebesgue-restricted'
  | 'gaussian-kernel'
  | 'product-measure'
  | 'custom';

export interface MinorizationData {
  kind: MinorizationKind;
  /** Minorization constant ε > 0 */
  epsilon: number;
  /** Number of steps for minorization (m-step) */
  steps: number;
  /** Lean expression for the minorizing measure */
  leanExpr: string;
}

/**
 * Handles the gnosis synthesize Minorization workflow.
 */
export function synthesizeMinorization(
  smallSet: SmallSet,
  properties: Record<string, string>
): MinorizationData {
  try {
    const epsilon = parseFloat(properties['minorization_epsilon'] ?? '0.01');
    const steps = parseInt(properties['minorization_steps'] ?? '1', 10);

    switch (smallSet.kind) {
      case 'queue-boundary':
        return {
          kind: 'uniform-on-small-set',
          epsilon,
          steps,
          leanExpr: `MeasureTheory.Measure.restrict volume ${smallSet.leanExpr}`,
        };

      case 'compact-sublevel':
        return {
          kind: 'lebesgue-restricted',
          epsilon,
          steps,
          leanExpr: `MeasureTheory.Measure.restrict volume ${smallSet.leanExpr}`,
        };

      case 'level-set':
        return {
          kind: 'dirac',
          epsilon,
          steps,
          leanExpr: `MeasureTheory.Measure.dirac x₀`,
        };

      default:
        return {
          kind: 'uniform-on-small-set',
          epsilon,
          steps,
          leanExpr: `${epsilon} • MeasureTheory.Measure.restrict volume ${smallSet.leanExpr}`,
        };
    }

  } catch (error) {
    throw new Error(`[synthesizeMinorization] ${error instanceof Error ? error.message : String(error)}`);
  }
}

// ============================================================================
// Harris certificate -- the full package
// ============================================================================

export interface HarrisCertificate {
  /** Node ID this certificate applies to */
  nodeId: string;
  /** Observable function */
  observable: Observable;
  /** Small set */
  smallSet: SmallSet;
  /** Lyapunov function */
  lyapunov: LyapunovFunction;
  /** Minorization data */
  minorization: MinorizationData;
  /** Whether all components are verified */
  fullyVerified: boolean;
  /** Generated theorem name */
  theoremName: string;
  /** Lean module name */
  moduleName: string;
}

/**
 * Handles the gnosis synthesize Harris Certificate workflow.
 */
export function synthesizeHarrisCertificate(
  nodeId: string,
  properties: Record<string, string>
): HarrisCertificate {
  const observable = inferObservable(properties);
  const smallSet = synthesizeSmallSet(nodeId, observable.kind, properties);
  const lyapunov = synthesizeLyapunov(observable, properties);
  const minorization = synthesizeMinorization(smallSet, properties);

  const normalizedId = nodeId.replace(/[^a-zA-Z0-9]/g, '_').toLowerCase();
  const theoremName = `${normalizedId}_harris_recurrent`;
  const moduleName = `Harris_${normalizedId}`;

  return {
    nodeId,
    observable,
    smallSet,
    lyapunov,
    minorization,
    fullyVerified: smallSet.verified,
    theoremName,
    moduleName,
  };
}

// ============================================================================
// Non-queue measurable kernels
// ============================================================================

export type KernelFamily =
  | 'queue-support'
  | 'diffusion'
  | 'jump-diffusion'
  | 'piecewise-deterministic'
  | 'renewal'
  | 'general-feller';

export interface MeasurableKernel {
  family: KernelFamily;
  /** State space description */
  stateSpace: string;
  /** Whether the kernel is Feller (continuous) */
  feller: boolean;
  /** Whether the kernel is strongly Feller */
  stronglyFeller: boolean;
  /** Lean type expression for the kernel */
  leanType: string;
}

/**
 * Handles the gnosis infer Kernel Family workflow.
 */
export function inferKernelFamily(
  properties: Record<string, string>
): MeasurableKernel {
  try {
    const family = (properties['kernel_family'] ??
      'queue-support') as KernelFamily;

    switch (family) {
      case 'diffusion':
        return {
          family,
          stateSpace: 'ℝ^n',
          feller: true,
          stronglyFeller: true,
          leanType:
            'MeasureTheory.Kernel (EuclideanSpace ℝ (Fin n)) (EuclideanSpace ℝ (Fin n))',
        };

      case 'jump-diffusion':
        return {
          family,
          stateSpace: 'ℝ^n',
          feller: true,
          stronglyFeller: false,
          leanType:
            'MeasureTheory.Kernel (EuclideanSpace ℝ (Fin n)) (EuclideanSpace ℝ (Fin n))',
        };

      case 'piecewise-deterministic':
        return {
          family,
          stateSpace: 'E × ℝ₊',
          feller: false,
          stronglyFeller: false,
          leanType: 'MeasureTheory.Kernel (E × ℝ≥0) (E × ℝ≥0)',
        };

      case 'renewal':
        return {
          family,
          stateSpace: 'ℕ × ℝ₊',
          feller: false,
          stronglyFeller: false,
          leanType: 'MeasureTheory.Kernel (ℕ × ℝ≥0) (ℕ × ℝ≥0)',
        };

      case 'general-feller':
        return {
          family,
          stateSpace: 'X (Polish)',
          feller: true,
          stronglyFeller: false,
          leanType: 'MeasureTheory.Kernel X X',
        };

      default:
        return {
          family: 'queue-support',
          stateSpace: 'ℕ',
          feller: false,
          stronglyFeller: false,
          leanType: 'MeasureTheory.Kernel ℕ ℕ',
        };
    }

  } catch (error) {
    throw new Error(`[inferKernelFamily] ${error instanceof Error ? error.message : String(error)}`);
  }
}

// ============================================================================
// Lean artifact generation for Harris certificates
// ============================================================================

/**
 * Handles the gnosis generate Harris Lean workflow.
 */
export function generateHarrisLean(cert: HarrisCertificate): string {
  const lines: string[] = [];
  lines.push(`/- Harris recurrence certificate for node '${cert.nodeId}' -/`);
  lines.push(`/- Generated by gnosis continuous-harris synthesis -/`);
  lines.push('');
  lines.push(`namespace ${cert.moduleName}`);
  lines.push('');
  lines.push(`/-- Observable: ${cert.observable.expression} -/`);
  lines.push(`noncomputable def observable := ${cert.observable.leanExpr}`);
  lines.push('');
  lines.push(`/-- Small set: ${cert.smallSet.description} -/`);
  lines.push(`def smallSet := ${cert.smallSet.leanExpr}`);
  lines.push('');
  lines.push(`/-- Lyapunov function: ${cert.lyapunov.expression} -/`);
  lines.push(`noncomputable def lyapunov := ${cert.lyapunov.leanExpr}`);
  lines.push('');
  lines.push(
    `/-- Drift condition: λ = ${cert.lyapunov.lambda}, b = ${cert.lyapunov.b} -/`
  );
  lines.push(`def driftLambda : ℝ := ${cert.lyapunov.lambda}`);
  lines.push(`def driftB : ℝ := ${cert.lyapunov.b}`);
  lines.push('');
  lines.push(
    `/-- Minorization: ε = ${cert.minorization.epsilon}, m = ${cert.minorization.steps} -/`
  );
  lines.push(`def minorizationEpsilon : ℝ := ${cert.minorization.epsilon}`);
  lines.push(`def minorizationSteps : ℕ := ${cert.minorization.steps}`);
  lines.push('');
  lines.push(
    `/-- ${cert.theoremName}: Harris recurrence with geometric ergodicity -/`
  );
  lines.push(`theorem ${cert.theoremName} :`);
  lines.push(
    `    HarrisRecurrent kernel smallSet minorizationEpsilon minorizationSteps`
  );
  lines.push(`    ∧ GeometricErgodic kernel lyapunov driftLambda driftB := by`);
  lines.push(`  exact ⟨harris_of_drift_and_minorization, geometric_of_drift⟩`);
  lines.push('');
  lines.push(`end ${cert.moduleName}`);

  return lines.join('\n');
}
