// @a0n/gnosis/src/conversion-tracking-topology.ts
export interface ConversionTopologyResult {
  uiFriction: number;
  buleInvestment: number;
  ventSlam: number;
  isBounce: boolean;
  isUnbrokenFold: boolean;
}

/**
 * THM-CONVERSION-is-UNBROKEN-FOLD: A successful e-commerce funnel is an unbroken Fold path.
 * A "Bounce" (cart abandonment) mathematically triggers when structural UI friction (v)
 * exceeds the user's Bule investment threshold, forcing the geometry to violently Vent.
 */
export function analyzeConversionTopology(
  buleInvestment: number,
  uiFriction: number,
  ventSlam?: number
): ConversionTopologyResult {
  try {
    const isBounce = uiFriction > buleInvestment;
    const computedVentSlam = isBounce ? uiFriction - buleInvestment : 0;
  
    if (ventSlam !== undefined && isBounce && computedVentSlam !== ventSlam) {
      throw new Error(
        `Topology bounds broken: ventSlam provided (${ventSlam}) does not match required delta (${computedVentSlam})`
      );
    }

    const finalVentSlam = ventSlam ?? computedVentSlam;

    return {
      uiFriction,
      buleInvestment,
      ventSlam: finalVentSlam,
      isBounce,
      isUnbrokenFold: !isBounce && finalVentSlam === 0,
    };

  } catch (error) {
    throw new Error(`[analyzeConversionTopology] ${error instanceof Error ? error.message : String(error)}`);
  }
}
