// @a0n/gnosis/src/cooling-policy.ts
/**
 * Load-aware cooling policy (TOPO_COGNITIVE_EVENT_HORIZON).
 * S_cog = clamp((K - L) / K). Machine-owned actuators: alert caps, cadence dilation.
 * Arousal proxy is Layer C — opaque numeric score from instrumentation only.
 */

/** Opaque: produced by external measurement pipeline (HRV, task errors, engineered score). */
export type ArousalProxy = number;

export interface CoolingPolicyParams {
  /** Max concurrent critical contexts (training/UI bound). */
  readonly K: number;
  /** Current active critical contexts L(t). */
  readonly L: number;
  /** Maximum alerts per minute before cooling forces throttle. */
  readonly alertRateMax: number;
}

export interface CoolingState {
  readonly arousalProxy: ArousalProxy;
  readonly alertRatePerMin: number;
  readonly mode: 'normal' | 'cooling';
}

/**
 * Handles the gnosis cognitive Slack workflow.
 */
export function cognitiveSlack(K: number, L: number): number {
  if (K <= 0) return 0;
  const raw = (K - L) / K;
  return Math.max(0, Math.min(1, raw));
}

/** Trigger cooling when slack below threshold or arousal/alert exceed bounds. */
export function coolingTrigger(
  params: CoolingPolicyParams,
  state: CoolingState,
  slackThreshold: number,
  arousalMax: number
): boolean {
  const slack = cognitiveSlack(params.K, params.L);
  return (
    slack < slackThreshold ||
    state.arousalProxy > arousalMax ||
    state.alertRatePerMin > params.alertRateMax
  );
}

/** Merge promotion allowed only if not in cooling (policy hook for Arena 5). */
export function mergeAllowedWhenCooling(state: CoolingState): boolean {
  return state.mode !== 'cooling';
}
