// @a0n/gnosis/src/cosmic-horizon.ts
/**
 * Cosmic Horizon Map -- classifies spacetime events relative to
 * cosmological horizons (particle, Hubble, event).
 *
 * All distances in light-years. All times in years.
 *
 * @packageDocumentation
 */

import type { SpacetimeEvent } from './causal-topology';
import {
  HUBBLE_CONSTANT_KM_S_MPC,
  MPC_IN_LY,
  SPEED_OF_LIGHT_MS,
  LIGHT_PICOLORENZO_LY,
} from './math';

// ─── Types ──────────────────────────────────────────────────────────

export interface CosmologicalParams {
  readonly hubbleConstantKmPerSecPerMpc: number;
  readonly matterDensity: number;
  readonly darkEnergyDensity: number;
  readonly ageOfUniverseYears: number;
}

export interface HorizonBody {
  readonly id: string;
  readonly distanceLy: number;
  readonly insideHorizon: boolean;
}

export interface CosmicHorizon {
  readonly source: 'GnosisCosmicHorizon';
  readonly cosmicTimeYears: number;
  readonly particleHorizonLy: number;
  readonly hubbleRadiusLy: number;
  readonly eventHorizonLy: number;
  readonly bodies: readonly HorizonBody[];
  readonly bodiesInsideCount: number;
  readonly bodiesOutsideCount: number;
  readonly horizonFraction: number;
}

// ─── Constants ──────────────────────────────────────────────────────

/** Meters per megaparsec. */
const METERS_PER_MPC = 3.0857e22;

/** Seconds per year. */
const SECONDS_PER_YEAR = 365.25 * 86400;

/** Present-day particle horizon radius in light-years. */
const PARTICLE_HORIZON_NOW_LY = 46.5e9;

/** Present-day age of the universe in years. */
const AGE_NOW = 13.8e9;

/** Future event horizon in a dark-energy-dominated universe (ly). */
const EVENT_HORIZON_LY = 16.5e9;

export const DEFAULT_COSMOLOGICAL_PARAMS: CosmologicalParams = {
  hubbleConstantKmPerSecPerMpc: HUBBLE_CONSTANT_KM_S_MPC,
  matterDensity: 0.315,
  darkEnergyDensity: 0.685,
  ageOfUniverseYears: AGE_NOW,
};

// ─── Functions ──────────────────────────────────────────────────────

/**
 * Hubble time 1/H_0 in years.
 *
 * H_0 in 1/s = H_0_km_s_Mpc * 1000 / METERS_PER_MPC.
 * Hubble time = 1 / H_0_per_s, converted to years.
 */
export function hubbleTimeYears(
  params: CosmologicalParams = DEFAULT_COSMOLOGICAL_PARAMS,
): number {
  const h0PerSecond =
    (params.hubbleConstantKmPerSecPerMpc * 1000) / METERS_PER_MPC;
  return 1 / h0PerSecond / SECONDS_PER_YEAR;
}

/**
 * Hubble radius c/H(t) in light-years.
 *
 * For simplification we assume H(t) ~ H_0 at the current epoch,
 * so hubbleRadius ~ hubbleTimeYears (since c = 1 ly/yr).
 */
export function hubbleRadius(
  cosmicTimeYears: number,
  params: CosmologicalParams = DEFAULT_COSMOLOGICAL_PARAMS,
): number {
  // c/H_0 in ly = hubbleTimeYears * (1 ly/yr) = hubbleTimeYears
  return hubbleTimeYears(params);
}

/**
 * Particle horizon radius at cosmic time t, in light-years.
 *
 * Uses the matter-dominated scaling: particleHorizon(t) ~ 46.5e9 * (t / 13.8e9)^(2/3).
 * At t = 0, returns 0. At t = 13.8e9, returns ~46.5e9 ly.
 */
export function particleHorizonRadius(
  cosmicTimeYears: number,
  params: CosmologicalParams = DEFAULT_COSMOLOGICAL_PARAMS,
): number {
  if (cosmicTimeYears <= 0) return 0;
  const fraction = cosmicTimeYears / params.ageOfUniverseYears;
  return PARTICLE_HORIZON_NOW_LY * Math.pow(fraction, 2 / 3);
}

/**
 * Future event horizon radius in light-years.
 *
 * In a dark-energy-dominated universe (Omega_Lambda ~ 0.685),
 * the event horizon converges to approximately 16.5 billion ly --
 * the maximum distance from which we will ever receive a signal.
 */
export function eventHorizonRadius(
  params: CosmologicalParams = DEFAULT_COSMOLOGICAL_PARAMS,
): number {
  return EVENT_HORIZON_LY;
}

/**
 * Euclidean distance of an event from the spatial origin, in light-years.
 */
function eventDistanceLy(event: SpacetimeEvent): number {
  return Math.sqrt(event.x * event.x + event.y * event.y + event.z * event.z);
}

/**
 * Build a cosmic horizon map for the given events.
 *
 * Each event is classified as inside or outside the particle horizon
 * at the given cosmic time. Default cosmic time maps to the present age
 * of the universe.
 */
export function cosmicHorizonMap(
  events: readonly SpacetimeEvent[],
  cosmicTimeYears?: number,
  params: CosmologicalParams = DEFAULT_COSMOLOGICAL_PARAMS,
): CosmicHorizon {
  const t = cosmicTimeYears ?? params.ageOfUniverseYears;
  const pHorizon = particleHorizonRadius(t, params);
  const hRadius = hubbleRadius(t, params);
  const eHorizon = eventHorizonRadius(params);

  const bodies: HorizonBody[] = events.map((ev) => {
    const dist = eventDistanceLy(ev);
    return {
      id: ev.id,
      distanceLy: dist,
      insideHorizon: dist <= pHorizon,
    };
  });

  const insideCount = bodies.filter((b) => b.insideHorizon).length;
  const outsideCount = bodies.length - insideCount;

  return {
    source: 'GnosisCosmicHorizon',
    cosmicTimeYears: t,
    particleHorizonLy: pHorizon,
    hubbleRadiusLy: hRadius,
    eventHorizonLy: eHorizon,
    bodies,
    bodiesInsideCount: insideCount,
    bodiesOutsideCount: outsideCount,
    horizonFraction: bodies.length > 0 ? insideCount / bodies.length : 0,
  };
}
