// @a0n/gnosis/src/cross-domain.ts
import { readFile } from 'node:fs/promises';
import path from 'node:path';
import {
  BettyCompiler,
  type ASTNode,
  type GraphAST,
} from './betty/compiler';
import type {
  SemanticContract,
  SemanticFacet,
  SemanticObligation,
  SemanticPredicate,
  TopologyType,
} from './betty/semantic-compatibility';
import {
  analyzePolyglotSource,
  type PolyglotAnalysisResult,
} from './polyglot-bridge';
import { analyzeCSSLayout } from './css-layout-analysis';
import { translate } from './polyglot-compose';
import { scaffoldTopology } from './polyglot-scaffold';

export type LanguageDomain = 'code' | 'natural' | 'gg';

export interface PreservationProfile {
  readonly required: readonly string[];
  readonly preferred: readonly string[];
}

export interface SourceArtifact {
  readonly domain: LanguageDomain;
  readonly language: string;
  readonly filePath?: string;
  readonly sourceText?: string;
  readonly format?: string;
  readonly metadata?: Readonly<Record<string, unknown>>;
}

export interface TargetArtifactDescriptor {
  readonly domain: LanguageDomain;
  readonly language: string;
  readonly format?: string;
}

export interface CrossDomainAnalysis {
  readonly adapterId: string;
  readonly source: SourceArtifact;
  readonly semanticContract: SemanticContract;
  readonly ggSource?: string;
  readonly ast?: GraphAST;
  readonly facets: readonly SemanticFacet[];
  readonly obligations: readonly SemanticObligation[];
  readonly diagnostics: readonly string[];
  readonly metadata?: Readonly<Record<string, unknown>>;
}

export interface TargetArtifact {
  readonly adapterId: string;
  readonly domain: LanguageDomain;
  readonly language: string;
  readonly text: string;
  readonly fileName?: string;
  readonly ggSource?: string;
  readonly facets: readonly SemanticFacet[];
  readonly obligations: readonly SemanticObligation[];
  readonly diagnostics: readonly string[];
  readonly metadata?: Readonly<Record<string, unknown>>;
}

export interface CrossCompileRequest {
  readonly source: SourceArtifact;
  readonly target: TargetArtifactDescriptor;
  readonly preservation?: Partial<PreservationProfile>;
  readonly registry?: LanguageAdapterRegistry;
}

export interface CrossCompileResult {
  readonly source: CrossDomainAnalysis;
  readonly target?: TargetArtifact;
  readonly diagnostics: readonly string[];
  readonly obligations: readonly SemanticObligation[];
  readonly success: boolean;
}

export interface CrossCompileContext {
  readonly registry: LanguageAdapterRegistry;
  readonly request: CrossCompileRequest;
}

export interface LanguageAdapter {
  readonly id: string;
  readonly domain: LanguageDomain;
  readonly languages: readonly string[];
  analyze?(
    source: SourceArtifact,
    context: CrossCompileContext
  ): Promise<CrossDomainAnalysis> | CrossDomainAnalysis;
  emit?(
    analysis: CrossDomainAnalysis,
    target: TargetArtifactDescriptor,
    context: CrossCompileContext
  ): Promise<TargetArtifact> | TargetArtifact;
  supportsSource?(source: SourceArtifact): boolean;
  supportsTarget?(target: TargetArtifactDescriptor): boolean;
}

const CODE_LANGUAGE_BY_EXTENSION: Record<string, string> = {
  '.ts': 'typescript',
  '.tsx': 'typescript',
  '.js': 'javascript',
  '.jsx': 'javascript',
  '.mjs': 'javascript',
  '.cjs': 'javascript',
  '.py': 'python',
  '.go': 'go',
  '.rs': 'rust',
  '.java': 'java',
  '.rb': 'ruby',
  '.php': 'php',
  '.lua': 'lua',
  '.swift': 'swift',
  '.kt': 'kotlin',
  '.scala': 'scala',
  '.hs': 'haskell',
  '.ml': 'ocaml',
  '.zig': 'zig',
  '.sh': 'bash',
  '.exs': 'elixir',
  '.cs': 'c_sharp',
  '.c': 'c',
  '.cpp': 'cpp',
  '.cc': 'cpp',
  '.cxx': 'cpp',
};

const BUILT_IN_CODE_EMIT_LANGUAGES = [
  'typescript',
  'javascript',
  'python',
  'go',
  'rust',
  'java',
  'ruby',
  'php',
  'lua',
  'swift',
  'kotlin',
  'scala',
  'haskell',
  'ocaml',
  'zig',
  'bash',
  'elixir',
  'c_sharp',
  'c',
  'cpp',
] as const;

const GG_LANGUAGE = 'gg';

export class LanguageAdapterRegistry {
  private readonly adapters = new Map<string, LanguageAdapter>();

  register(adapter: LanguageAdapter): void {
    this.adapters.set(adapter.id, adapter);
  }

  unregister(adapterId: string): void {
    this.adapters.delete(adapterId);
  }

  list(): LanguageAdapter[] {
    return [...this.adapters.values()];
  }

  resolveAnalyzer(source: SourceArtifact): LanguageAdapter | undefined {
    return this.list().find((adapter) => {
      if (adapter.domain !== source.domain || !adapter.analyze) {
        return false;
      }
      if (adapter.supportsSource) {
        return adapter.supportsSource(source);
      }
      return adapter.languages.includes(source.language);
    });
  }

  resolveEmitter(
    target: TargetArtifactDescriptor
  ): LanguageAdapter | undefined {
    return this.list().find((adapter) => {
      if (adapter.domain !== target.domain || !adapter.emit) {
        return false;
      }
      if (adapter.supportsTarget) {
        return adapter.supportsTarget(target);
      }
      return adapter.languages.includes(target.language);
    });
  }
}

export const defaultLanguageAdapterRegistry = new LanguageAdapterRegistry();

/**
 * Creates the Preservation Profile.
 */
export function createPreservationProfile(
  profile?: Partial<PreservationProfile>
): PreservationProfile {
  return {
    required: [...(profile?.required ?? [])],
    preferred: [...(profile?.preferred ?? [])],
  };
}

/**
 * Handles the gnosis infer Code Language From File Path workflow.
 */
export function inferCodeLanguageFromFilePath(
  filePath: string
): string | undefined {
  const ext = path.extname(filePath).toLowerCase();
  return CODE_LANGUAGE_BY_EXTENSION[ext];
}

/**
 * Handles the gnosis infer Language Domain From File Path workflow.
 */
export function inferLanguageDomainFromFilePath(
  filePath: string
): LanguageDomain | undefined {
  try {
    const ext = path.extname(filePath).toLowerCase();
    if (ext === '.gg' || ext === '.mgg') {
      return 'gg';
    }
    if (CODE_LANGUAGE_BY_EXTENSION[ext]) {
      return 'code';
    }
    return undefined;

  } catch (error) {
    throw new Error(`[inferLanguageDomainFromFilePath] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Handles the gnosis cross Compile workflow.
 */
export async function crossCompile(
  request: CrossCompileRequest
): Promise<CrossCompileResult> {
  try {
    const registry = request.registry ?? defaultLanguageAdapterRegistry;
    const preservation = createPreservationProfile(request.preservation);
    const context: CrossCompileContext = {
      registry,
      request: {
        ...request,
        preservation,
      },
    };

    const analyzer = registry.resolveAnalyzer(request.source);
    if (!analyzer?.analyze) {
      return failureResult(request.source, [
        createObligation(
          'missing-source-adapter',
          `No source adapter is registered for ${request.source.domain}:${request.source.language}`,
          'error',
          true
        ),
      ]);
    }

    const sourceAnalysis = await analyzer.analyze(request.source, context);
    const emitter = registry.resolveEmitter(request.target);
    if (!emitter?.emit) {
      return {
        source: sourceAnalysis,
        diagnostics: sourceAnalysis.diagnostics,
        obligations: [
          ...sourceAnalysis.obligations,
          createObligation(
            'missing-target-adapter',
            `No target adapter is registered for ${request.target.domain}:${request.target.language}`,
            'error',
            true
          ),
        ],
        success: false,
      };
    }

    const targetArtifact = await emitter.emit(
      sourceAnalysis,
      request.target,
      context
    );
    const preservationObligations = validatePreservation(
      preservation,
      sourceAnalysis.facets,
      targetArtifact.facets
    );
    const obligations = dedupeObligations([
      ...sourceAnalysis.obligations,
      ...targetArtifact.obligations,
      ...preservationObligations,
    ]);
    const diagnostics = [
      ...sourceAnalysis.diagnostics,
      ...targetArtifact.diagnostics,
    ];

    return {
      source: sourceAnalysis,
      target: targetArtifact,
      diagnostics,
      obligations,
      success: obligations.every((obligation) => obligation.severity !== 'error'),
    };

  } catch (error) {
    throw new Error(`[crossCompile] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Handles the gnosis install Built In Language Adapters workflow.
 */
export function installBuiltInLanguageAdapters(
  registry: LanguageAdapterRegistry = defaultLanguageAdapterRegistry
): LanguageAdapterRegistry {
  registry.register(codeLanguageAdapter);
  registry.register(ggLanguageAdapter);
  return registry;
}

async function readSourceArtifactText(source: SourceArtifact): Promise<string> {
  try {
    if (typeof source.sourceText === 'string') {
      return source.sourceText;
    }
    if (!source.filePath) {
      throw new Error(
        `source text is required for ${source.domain}:${source.language} artifacts`
      );
    }
    return readFile(source.filePath, 'utf8');

  } catch (error) {
    throw new Error(`[readSourceArtifactText] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function createEmptyContract(): SemanticContract {
  return {
    paramTypes: [],
    returnType: { kind: 'unknown' },
    predicates: [],
    facets: [],
    obligations: [],
  };
}

function normalizeSemanticContract(
  contract?: Partial<{
    param_types: TopologyType[];
    return_type: TopologyType;
    predicates: SemanticPredicate[];
    facets: SemanticFacet[];
    obligations: SemanticObligation[];
    paramTypes: TopologyType[];
    returnType: TopologyType;
  }>
): SemanticContract {
  return {
    paramTypes: [...(contract?.paramTypes ?? contract?.param_types ?? [])],
    returnType: contract?.returnType ??
      contract?.return_type ?? { kind: 'unknown' },
    predicates: [...(contract?.predicates ?? [])],
    facets: [...(contract?.facets ?? [])],
    obligations: [...(contract?.obligations ?? [])],
  };
}

function createObligation(
  key: string,
  message: string,
  severity: SemanticObligation['severity'],
  required = false,
  facetKey?: string,
  resolution?: string
): SemanticObligation {
  const obligation: SemanticObligation = {
    key,
    message,
    severity,
    required,
  };
  if (facetKey !== undefined) {
    obligation.facetKey = facetKey;
  }
  if (resolution !== undefined) {
    obligation.resolution = resolution;
  }
  return obligation;
}

function dedupeFacets(facets: readonly SemanticFacet[]): SemanticFacet[] {
  try {
    const seen = new Set<string>();
    const result: SemanticFacet[] = [];
    for (const facet of facets) {
      const fingerprint = `${facet.key}\u0000${facet.value}\u0000${
        facet.domain ?? ''
      }\u0000${facet.detail ?? ''}`;
      if (seen.has(fingerprint)) {
        continue;
      }
      seen.add(fingerprint);
      result.push(facet);
    }
    return result;

  } catch (error) {
    throw new Error(`[dedupeFacets] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function dedupeObligations(
  obligations: readonly SemanticObligation[]
): SemanticObligation[] {
  try {
    const seen = new Set<string>();
    const result: SemanticObligation[] = [];
    for (const obligation of obligations) {
      const fingerprint = `${obligation.key}\u0000${obligation.message}\u0000${obligation.severity}`;
      if (seen.has(fingerprint)) {
        continue;
      }
      seen.add(fingerprint);
      result.push(obligation);
    }
    return result;

  } catch (error) {
    throw new Error(`[dedupeObligations] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function findEntryNode(ast: GraphAST): ASTNode | undefined {
  return [...ast.nodes.values()].find((node) => node.labels.includes('Entry'));
}

function parseStructuredProperty<T>(
  rawValue: string | undefined,
  encodedValue: string | undefined,
  fallback: T
): T {
  const serialized =
    typeof encodedValue === 'string'
      ? Buffer.from(encodedValue, 'base64').toString('utf8')
      : rawValue;

  if (!serialized) {
    return fallback;
  }

  try {
    return JSON.parse(serialized) as T;
  } catch {
    return fallback;
  }
}

function extractContractFromNode(node: ASTNode | undefined): SemanticContract {
  try {
    if (!node) {
      return createEmptyContract();
    }

    const contract = createEmptyContract();
    const semanticReturn = node.properties['semantic_return_type'];
    const semanticParams = node.properties['semantic_param_types'];
    const semanticPredicates = node.properties['semantic_predicates'];
    const semanticFacets = node.properties['semantic_facets'];
    const semanticObligations = node.properties['semantic_obligations'];
    const semanticReturnEncoded = node.properties['semantic_return_type_b64'];
    const semanticParamsEncoded = node.properties['semantic_param_types_b64'];
    const semanticPredicatesEncoded = node.properties['semantic_predicates_b64'];
    const semanticFacetsEncoded = node.properties['semantic_facets_b64'];
    const semanticObligationsEncoded =
      node.properties['semantic_obligations_b64'];

    contract.returnType = parseStructuredProperty(
      semanticReturn,
      semanticReturnEncoded,
      { kind: 'unknown' } satisfies TopologyType
    );
    contract.paramTypes = parseStructuredProperty(
      semanticParams,
      semanticParamsEncoded,
      [] as TopologyType[]
    );
    contract.predicates = parseStructuredProperty(
      semanticPredicates,
      semanticPredicatesEncoded,
      [] as SemanticPredicate[]
    );
    contract.facets = parseStructuredProperty(
      semanticFacets,
      semanticFacetsEncoded,
      [] as SemanticFacet[]
    );
    contract.obligations = parseStructuredProperty(
      semanticObligations,
      semanticObligationsEncoded,
      [] as SemanticObligation[]
    );

    return contract;

  } catch (error) {
    throw new Error(`[extractContractFromNode] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function deriveCodeFacets(analysis: PolyglotAnalysisResult): SemanticFacet[] {
  try {
    const facets: SemanticFacet[] = [];
    const signatures = analysis.functions
      .map((func) => func.signature)
      .filter((signature): signature is NonNullable<typeof signature> =>
        Boolean(signature)
      );

    if (signatures.some((signature) => signature.is_async)) {
      facets.push({
        key: 'async',
        value: 'present',
        domain: 'code',
        detail: 'Source contains async control flow.',
        provenance: analysis.language,
      });
    }

    if (
      signatures.some((signature) => {
        const returnType = signature.return_type ?? '';
        return /(Result<|error\b|Exception\b|throws\b)/i.test(returnType);
      })
    ) {
      facets.push({
        key: 'error-discipline',
        value: 'explicit',
        domain: 'code',
        detail: 'Source encodes explicit error handling in its surface types.',
        provenance: analysis.language,
      });
    }

    if (
      analysis.language === 'rust' &&
      signatures.some((signature) =>
        signature.params.some((param) =>
          /(&|mut\b|Box<|Arc<|Rc<)/.test(param.type_annotation ?? '')
        )
      )
    ) {
      facets.push({
        key: 'ownership',
        value: 'explicit',
        domain: 'code',
        detail: 'Rust ownership and borrowing markers are present.',
        provenance: 'rust',
      });
    }

    return dedupeFacets(facets);

  } catch (error) {
    throw new Error(`[deriveCodeFacets] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function createFunctionNameFromAnalysis(analysis: CrossDomainAnalysis): string {
  try {
    const fromMetadata = analysis.metadata?.['suggestedFunctionName'];
    if (typeof fromMetadata === 'string' && fromMetadata.length > 0) {
      return fromMetadata;
    }
    const entryNode = analysis.ast ? findEntryNode(analysis.ast) : undefined;
    const fromEntry = entryNode?.properties.name;
    if (typeof fromEntry === 'string' && fromEntry.length > 0) {
      return fromEntry;
    }
    return 'cross_domain_flow';

  } catch (error) {
    throw new Error(`[createFunctionNameFromAnalysis] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function summarizeFacets(facets: readonly SemanticFacet[]): string {
  try {
    if (facets.length === 0) {
      return 'none';
    }
    return facets.map((facet) => `${facet.key}=${facet.value}`).join(', ');

  } catch (error) {
    throw new Error(`[summarizeFacets] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function buildCodeToNaturalSummary(
  analysis: CrossDomainAnalysis,
  targetLanguage: string
): string {
  try {
    const polyglotAnalysis = analysis.metadata?.['polyglotAnalysis'] as
      | PolyglotAnalysisResult
      | undefined;
    const functionNames =
      polyglotAnalysis?.functions.map((func) => func.functionName) ?? [];
    const functionSummary =
      functionNames.length > 0
        ? functionNames.join(', ')
        : createFunctionNameFromAnalysis(analysis);
    const nuanceSummary = summarizeFacets(analysis.facets);

    switch (targetLanguage) {
      case 'es':
        return `Este artefacto describe ${functionSummary}. Matices preservados: ${nuanceSummary}.`;
      case 'zh':
        return `这个制品描述了 ${functionSummary}。已保留的细节：${nuanceSummary}。`;
      default:
        return `This artifact describes ${functionSummary}. Preserved nuances: ${nuanceSummary}.`;
    }

  } catch (error) {
    throw new Error(`[buildCodeToNaturalSummary] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function validatePreservation(
  profile: PreservationProfile,
  sourceFacets: readonly SemanticFacet[],
  targetFacets: readonly SemanticFacet[]
): SemanticObligation[] {
  try {
    const obligations: SemanticObligation[] = [];
    const sourceFacetKeys = new Set(sourceFacets.map((facet) => facet.key));
    const targetFacetKeys = new Set(targetFacets.map((facet) => facet.key));

    for (const requiredFacet of profile.required) {
      if (!sourceFacetKeys.has(requiredFacet)) {
        continue;
      }
      if (!targetFacetKeys.has(requiredFacet)) {
        obligations.push(
          createObligation(
            `preserve-${requiredFacet}`,
            `Target artifact does not preserve required facet '${requiredFacet}'.`,
            'error',
            true,
            requiredFacet,
            'Choose a richer target or relax the preservation profile.'
          )
        );
      }
    }

    for (const preferredFacet of profile.preferred) {
      if (
        !sourceFacetKeys.has(preferredFacet) ||
        targetFacetKeys.has(preferredFacet)
      ) {
        continue;
      }
      obligations.push(
        createObligation(
          `prefer-${preferredFacet}`,
          `Target artifact dropped preferred facet '${preferredFacet}'.`,
          'warning',
          false,
          preferredFacet,
          'Prefer a target adapter that can surface the nuance explicitly.'
        )
      );
    }

    return obligations;

  } catch (error) {
    throw new Error(`[validatePreservation] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function failureResult(
  source: SourceArtifact,
  obligations: SemanticObligation[]
): CrossCompileResult {
  return {
    source: {
      adapterId: 'gnosis:error',
      source,
      semanticContract: createEmptyContract(),
      facets: [],
      obligations,
      diagnostics: obligations.map((obligation) => obligation.message),
    },
    diagnostics: obligations.map((obligation) => obligation.message),
    obligations,
    success: false,
  };
}

const codeLanguageAdapter: LanguageAdapter = {
  id: 'gnosis:code',
  domain: 'code',
  languages: BUILT_IN_CODE_EMIT_LANGUAGES,
  async analyze(source): Promise<CrossDomainAnalysis> {
    try {
      if (!source.filePath) {
        throw new Error('code analysis requires a file path');
      }

      const polyglotAnalysis = await analyzePolyglotSource(source.filePath);
      const primaryFunction = polyglotAnalysis.functions[0];
      const contract = normalizeSemanticContract(
        primaryFunction?.signature?.semantic_contract as
          | Partial<{
              param_types: TopologyType[];
              return_type: TopologyType;
              predicates: SemanticPredicate[];
              facets: SemanticFacet[];
              obligations: SemanticObligation[];
            }>
          | undefined
      );
      // CSS Layout Thermodynamics analysis for TSX/JSX/TS/JS files.
      // Runs the seven layout detectors backed by Lean theorems from
      // CSSLayoutThermodynamics.lean and surfaces findings as obligations.
      const sourceText =
        source.sourceText ?? (await readSourceArtifactText(source));
      const cssLayoutResult = analyzeCSSLayout(sourceText, source.filePath);

      const derivedFacets = dedupeFacets([
        ...contract.facets,
        ...deriveCodeFacets(polyglotAnalysis),
        ...(cssLayoutResult.findings.length > 0
          ? [
              {
                key: 'css-layout-thermodynamics',
                value: `${cssLayoutResult.findings.length} findings`,
                domain: 'layout' as const,
                detail: `CSS layout analysis detected ${cssLayoutResult.findings.length} issue(s) backed by Lean theorems.`,
                provenance: 'gnosis:css-layout',
              },
            ]
          : []),
      ]);
      const contractWithFacets: SemanticContract = {
        ...contract,
        facets: derivedFacets,
        obligations: dedupeObligations([
          ...contract.obligations,
          ...cssLayoutResult.obligations,
        ]),
      };

      return {
        adapterId: this.id,
        source: {
          ...source,
          sourceText,
        },
        semanticContract: contractWithFacets,
        ggSource: polyglotAnalysis.functions
          .map((func) => func.ggSource)
          .join('\n\n'),
        facets: derivedFacets,
        obligations: contractWithFacets.obligations,
        diagnostics: [
          ...polyglotAnalysis.errors,
          ...cssLayoutResult.diagnostics.map(
            (d) => `[${d.code ?? 'LAYOUT'}] L${d.line}: ${d.message}`
          ),
        ],
        metadata: {
          polyglotAnalysis,
          functionCount: polyglotAnalysis.functions.length,
          suggestedFunctionName: primaryFunction?.functionName,
        },
        ...(primaryFunction?.ast !== undefined
          ? { ast: primaryFunction.ast }
          : {}),
      };
  
    } catch (error) {
      throw new Error(`[analyze] ${error instanceof Error ? error.message : String(error)}`);
    }
},
  async emit(analysis, target): Promise<TargetArtifact> {
    try {
      if (target.language === GG_LANGUAGE) {
        return {
          adapterId: this.id,
          domain: 'code',
          language: target.language,
          text: analysis.ggSource ?? '',
          facets: analysis.facets,
          obligations: analysis.obligations,
          diagnostics: analysis.diagnostics,
          ...(analysis.ggSource !== undefined
            ? { ggSource: analysis.ggSource }
            : {}),
        };
      }

      const sourceFilePath = analysis.source.filePath;
      const polyglotAnalysis = analysis.metadata?.['polyglotAnalysis'] as
        | PolyglotAnalysisResult
        | undefined;

      if (
        analysis.source.domain === 'code' &&
        polyglotAnalysis &&
        sourceFilePath
      ) {
        const translated = translate(
          polyglotAnalysis,
          sourceFilePath,
          target.language
        );
        const text = translated.files.map((file) => file.source).join('\n\n');

        return {
          adapterId: this.id,
          domain: 'code',
          language: target.language,
          text,
          facets: analysis.facets,
          obligations: analysis.obligations,
          diagnostics: analysis.diagnostics,
          metadata: {
            fileCount: translated.files.length,
          },
          ...(translated.files[0]?.fileName !== undefined
            ? { fileName: translated.files[0].fileName }
            : {}),
          ...(translated.ggSource !== undefined
            ? { ggSource: translated.ggSource }
            : {}),
        };
      }

      if (!analysis.ast) {
        const artifact: TargetArtifact = {
          adapterId: this.id,
          domain: 'code',
          language: target.language,
          text: '',
          facets: analysis.facets,
          obligations: [
            ...analysis.obligations,
            createObligation(
              'missing-ast',
              `Cannot emit ${target.language} without a GG/AST intermediate.`,
              'error',
              true
            ),
          ],
          diagnostics: [
            ...analysis.diagnostics,
            `Missing AST for ${analysis.source.domain}:${analysis.source.language} -> code emission.`,
          ],
        };
        return artifact;
      }

      const entryNode =
        findEntryNode(analysis.ast) ?? [...analysis.ast.nodes.values()][0];
      if (!entryNode) {
        const artifact: TargetArtifact = {
          adapterId: this.id,
          domain: 'code',
          language: target.language,
          text: '',
          facets: analysis.facets,
          obligations: [
            ...analysis.obligations,
            createObligation(
              'empty-topology',
              `Cannot emit ${target.language} from an empty topology.`,
              'error',
              true
            ),
          ],
          diagnostics: analysis.diagnostics,
        };
        return artifact;
      }

      const scaffold = scaffoldTopology(
        analysis.ast,
        [
          {
            nodeId: entryNode.id,
            language: target.language,
            functionName: createFunctionNameFromAnalysis(analysis),
          },
        ],
        analysis.ggSource
      );

      return {
        adapterId: this.id,
        domain: 'code',
        language: target.language,
        text: scaffold.files.map((file) => file.source).join('\n\n'),
        facets: analysis.facets,
        obligations: analysis.obligations,
        diagnostics: analysis.diagnostics,
        ...(scaffold.files[0]?.fileName !== undefined
          ? { fileName: scaffold.files[0].fileName }
          : {}),
        ...(scaffold.annotatedTopology !== undefined
          ? { ggSource: scaffold.annotatedTopology }
          : {}),
      };
  
    } catch (error) {
      throw new Error(`[emit] ${error instanceof Error ? error.message : String(error)}`);
    }
},
};

const ggLanguageAdapter: LanguageAdapter = {
  id: 'gnosis:gg',
  domain: 'gg',
  languages: [GG_LANGUAGE],
  async analyze(source): Promise<CrossDomainAnalysis> {
    const sourceText = await readSourceArtifactText(source);
    const compiler = new BettyCompiler();
    const parsed = compiler.parse(sourceText);
    const ast = parsed.ast ?? { nodes: new Map<string, ASTNode>(), edges: [] };
    const contract = extractContractFromNode(findEntryNode(ast));

    return {
      adapterId: this.id,
      source: {
        ...source,
        sourceText,
      },
      semanticContract: contract,
      ggSource: sourceText,
      ast,
      facets: contract.facets,
      obligations: contract.obligations,
      diagnostics: parsed.diagnostics.map((diagnostic) => diagnostic.message),
    };
  },
  async emit(analysis, target): Promise<TargetArtifact> {
    const text = analysis.ggSource ?? '';
    return {
      adapterId: this.id,
      domain: target.domain,
      language: target.language,
      text,
      ggSource: text,
      facets: analysis.facets,
      obligations: analysis.obligations,
      diagnostics: analysis.diagnostics,
    };
  },
};

installBuiltInLanguageAdapters();

/**
 * Creates the Source Artifact From File.
 */
export function createSourceArtifactFromFile(
  filePath: string,
  overrides?: Partial<Omit<SourceArtifact, 'filePath'>>
): SourceArtifact {
  try {
    const inferredDomain =
      overrides?.domain ?? inferLanguageDomainFromFilePath(filePath);
    const inferredLanguage =
      overrides?.language ??
      (inferredDomain === 'code'
        ? inferCodeLanguageFromFilePath(filePath)
        : inferredDomain === 'gg'
        ? GG_LANGUAGE
        : undefined);

    if (!inferredDomain || !inferredLanguage) {
      throw new Error(
        `Unable to infer source artifact metadata from ${filePath}.`
      );
    }

    return {
      domain: inferredDomain,
      language: inferredLanguage,
      filePath,
      ...(overrides?.sourceText !== undefined
        ? { sourceText: overrides.sourceText }
        : {}),
      ...(overrides?.format !== undefined ? { format: overrides.format } : {}),
      ...(overrides?.metadata !== undefined
        ? { metadata: overrides.metadata }
        : {}),
    };

  } catch (error) {
    throw new Error(`[createSourceArtifactFromFile] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Builds the Code To Natural Artifact.
 */
export function buildCodeToNaturalArtifact(
  analysis: CrossDomainAnalysis,
  targetLanguage: string
): TargetArtifact {
  return {
    adapterId: 'gnosis:code-to-natural',
    domain: 'natural',
    language: targetLanguage,
    text: buildCodeToNaturalSummary(analysis, targetLanguage),
    facets: analysis.facets,
    obligations: analysis.obligations,
    diagnostics: analysis.diagnostics,
  };
}
