// @a0n/gnosis/src/daylight-topology.ts
/**
 * Daylight Topology — gradient-based daylight/twilight/night visualization.
 *
 * Produces immutable topology packets mapping time spans to twilight-zone
 * gradient stops, contiguous bands, and 2D spacetime surfaces. Consumed by
 * both stars.fish (astronomical context) and storms.watch (weather context).
 *
 * Two key features:
 *   1. `isDaylight(coordinate)` — O(1) daylight check for any atlas/mesh coordinate
 *   2. `createDaylightMesh()` — quantized daylight grid aligned to aeon-weather mesh
 *
 * @packageDocumentation
 */

import {
  toPicolorenzos,
  formatGnosticTime,
  PICOLORENZO_MS,
  GNOSTIC,
} from './math';

import type { TwilightZone, SolarObserver, SolarPosition } from './math';
import {
  solarPosition,
  classifyTwilightZone,
  twilightZoneFraction,
  solarPhase,
  solarPhaseContinuous,
  solarHalfDay,
} from './math';

export type { TwilightZone, SolarObserver, SolarPosition };
export { solarPhase, solarPhaseContinuous, solarHalfDay };

// ─── Color Types ────────────────────────────────────────────────────

export interface DaylightColor {
  readonly r: number;
  readonly g: number;
  readonly b: number;
  readonly a: number;
}

export interface DaylightPalette {
  readonly day: DaylightColor;
  readonly civilTwilight: DaylightColor;
  readonly nauticalTwilight: DaylightColor;
  readonly astronomicalTwilight: DaylightColor;
  readonly night: DaylightColor;
}

// ─── Default Palette (dark-theme compatible) ────────────────────────

export const DEFAULT_DAYLIGHT_PALETTE: DaylightPalette = {
  day:                  { r: 135, g: 206, b: 235, a: 1.0 },  // sky blue
  civilTwilight:        { r: 255, g: 140, b:  66, a: 1.0 },  // warm amber
  nauticalTwilight:     { r: 102, g:  51, b: 153, a: 1.0 },  // deep purple
  astronomicalTwilight: { r:  25, g:  25, b:  80, a: 1.0 },  // dark navy
  night:                { r:  13, g:  15, b:  24, a: 1.0 },  // void black
};

// ─── Gradient Stop ──────────────────────────────────────────────────

export interface DaylightGradientStop {
  readonly position: number;       // 0-1 on time axis
  readonly epochMs: number;
  readonly pLo: number;
  readonly altitudeDeg: number;
  readonly zone: TwilightZone;
  readonly zoneFraction: number;
  readonly color: DaylightColor;
  readonly gnosticTime: string;
}

// ─── Band ───────────────────────────────────────────────────────────

export interface DaylightBand {
  readonly zone: TwilightZone;
  readonly startPosition: number;
  readonly endPosition: number;
  readonly durationMs: number;
  readonly gradientStops: readonly DaylightGradientStop[];
}

// ─── Summary ────────────────────────────────────────────────────────

export interface DaylightSummary {
  readonly daylightFraction: number;
  readonly nightFraction: number;
  readonly civilTwilightFraction: number;
  readonly nauticalTwilightFraction: number;
  readonly astronomicalTwilightFraction: number;
}

// ─── Topology Packet ────────────────────────────────────────────────

export interface DaylightTopologyPacket {
  readonly source: 'GnosisDaylightTopology';
  readonly observer: SolarObserver;
  readonly startEpochMs: number;
  readonly endEpochMs: number;
  readonly spanPLo: number;
  readonly resolution: number;
  readonly bands: readonly DaylightBand[];
  readonly gradientStops: readonly DaylightGradientStop[];
  readonly summary: DaylightSummary;
}

// ─── Spacetime Surface ──────────────────────────────────────────────

export interface SpacetimeCell {
  readonly longitudeDeg: number;
  readonly epochMs: number;
  readonly altitudeDeg: number;
  readonly zone: TwilightZone;
  readonly color: DaylightColor;
}

export interface SpacetimeSurface {
  readonly source: 'GnosisDaylightSpacetime';
  readonly longitudeBins: number;
  readonly timeBins: number;
  readonly startEpochMs: number;
  readonly endEpochMs: number;
  readonly latitudeDeg: number;
  readonly cells: readonly SpacetimeCell[];  // row-major: [time][longitude]
}

// ─── Daylight Mesh (quantized grid) ─────────────────────────────────

export interface DaylightMeshCell {
  readonly latIndex: number;
  readonly lonIndex: number;
  readonly centroidLat: number;
  readonly centroidLon: number;
  readonly altitudeDeg: number;
  readonly zone: TwilightZone;
  readonly isDaylight: boolean;
  readonly color: DaylightColor;
}

export interface DaylightMeshPacket {
  readonly source: 'GnosisDaylightMesh';
  readonly epochMs: number;
  readonly resolutionDeg: number;
  readonly latCount: number;
  readonly lonCount: number;
  readonly cells: readonly DaylightMeshCell[];
  readonly daylightFraction: number;
}

// ─── Universal Coordinate Interface ─────────────────────────────────

/**
 * Minimal coordinate contract that both atlas coordinates and mesh
 * coordinates satisfy. Any object with lat, lon, and a time can be
 * checked for daylight.
 */
export interface DaylightCoordinate {
  readonly lat: number;
  readonly lon: number;
  readonly epochMs: number;
}

// ─── Color Interpolation ────────────────────────────────────────────

function lerpColor(a: DaylightColor, b: DaylightColor, t: number): DaylightColor {
  const ct = Math.max(0, Math.min(1, t));
  return {
    r: Math.round(a.r + (b.r - a.r) * ct),
    g: Math.round(a.g + (b.g - a.g) * ct),
    b: Math.round(a.b + (b.b - a.b) * ct),
    a: a.a + (b.a - a.a) * ct,
  };
}

const ZONE_ORDER: readonly TwilightZone[] = [
  'night',
  'astronomical',
  'nautical',
  'civil',
  'day',
];

function paletteForZone(palette: DaylightPalette, zone: TwilightZone): DaylightColor {
  switch (zone) {
    case 'day': return palette.day;
    case 'civil': return palette.civilTwilight;
    case 'nautical': return palette.nauticalTwilight;
    case 'astronomical': return palette.astronomicalTwilight;
    case 'night': return palette.night;
  }
}

/**
 * Interpolate a daylight color from solar altitude.
 * Smoothly lerps between adjacent zone colors using the zone fraction.
 */
export function interpolateDaylightColor(
  palette: DaylightPalette,
  altitudeDeg: number
): DaylightColor {
  const zone = classifyTwilightZone(altitudeDeg);
  const fraction = twilightZoneFraction(altitudeDeg);
  const zoneIdx = ZONE_ORDER.indexOf(zone);

  // At the upper boundary, lerp toward the next-higher zone
  if (zoneIdx < ZONE_ORDER.length - 1 && fraction > 0.5) {
    const nextZone = ZONE_ORDER[zoneIdx + 1]!;
    const t = (fraction - 0.5) * 2; // 0-1 in transition region
    return lerpColor(
      paletteForZone(palette, zone),
      paletteForZone(palette, nextZone),
      t
    );
  }

  // At the lower boundary, lerp toward the next-lower zone
  if (zoneIdx > 0 && fraction < 0.5) {
    const prevZone = ZONE_ORDER[zoneIdx - 1]!;
    const t = (0.5 - fraction) * 2; // 0-1 in transition region
    return lerpColor(
      paletteForZone(palette, zone),
      paletteForZone(palette, prevZone),
      t
    );
  }

  return paletteForZone(palette, zone);
}

function colorToRgba(c: DaylightColor): string {
  return `rgba(${c.r},${c.g},${c.b},${c.a})`;
}

// ─── isDaylight ─────────────────────────────────────────────────────

/**
 * O(1) daylight check for any coordinate with lat, lon, and time.
 *
 * Works with atlas coordinates, mesh coordinates, or any object that
 * provides `{ lat, lon, epochMs }`.
 *
 * Returns true if the sun is above the horizon (altitude > 0).
 */
export function isDaylight(coordinate: DaylightCoordinate): boolean {
  const result = solarPosition(
    { latitudeDeg: coordinate.lat, longitudeDeg: coordinate.lon },
    coordinate.epochMs
  );
  return result.zone === 'day';
}

/**
 * Full daylight state for any coordinate. Returns altitude, zone,
 * color, and daylight boolean.
 */
export function daylightState(
  coordinate: DaylightCoordinate,
  palette: DaylightPalette = DEFAULT_DAYLIGHT_PALETTE
): {
  readonly isDaylight: boolean;
  readonly altitude: SolarPosition;
  readonly color: DaylightColor;
} {
  const altitude = solarPosition(
    { latitudeDeg: coordinate.lat, longitudeDeg: coordinate.lon },
    coordinate.epochMs
  );
  return {
    isDaylight: altitude.zone === 'day',
    altitude,
    color: interpolateDaylightColor(palette, altitude.altitudeDeg),
  };
}

// ─── Topology Packet Construction ───────────────────────────────────

export interface CreateDaylightTopologyInput {
  readonly observer: SolarObserver;
  readonly startEpochMs?: number;
  readonly spanPLo?: number;
  readonly resolution?: number;
  readonly palette?: DaylightPalette;
}

/**
 * Create a daylight topology packet.
 *
 * Samples solar altitude at `resolution` evenly-spaced stops across
 * `spanPLo` picolorenzos starting from `startEpochMs`. Groups stops
 * into contiguous twilight-zone bands and computes summary fractions.
 *
 * Defaults:
 *   - span: 1 pLo (~3.14 days, ~3 day/night cycles)
 *   - resolution: 55 (Pleroma -- one sample per ~82 minutes)
 *   - palette: DEFAULT_DAYLIGHT_PALETTE
 */
export function createDaylightTopology(
  input: CreateDaylightTopologyInput
): DaylightTopologyPacket {
  const {
    observer,
    startEpochMs = Date.now(),
    spanPLo = 1,
    resolution = GNOSTIC.PLEROMA,
    palette = DEFAULT_DAYLIGHT_PALETTE,
  } = input;

  const totalMs = spanPLo * PICOLORENZO_MS;
  const endEpochMs = startEpochMs + totalMs;
  const stepMs = totalMs / Math.max(1, resolution - 1);

  // Generate gradient stops
  const stops: DaylightGradientStop[] = [];
  for (let i = 0; i < resolution; i++) {
    const epochMs = startEpochMs + i * stepMs;
    const position = i / Math.max(1, resolution - 1);
    const alt = solarPosition(observer, epochMs);
    const color = interpolateDaylightColor(palette, alt.altitudeDeg);
    const pLo = toPicolorenzos(new Date(epochMs));

    stops.push({
      position,
      epochMs,
      pLo,
      altitudeDeg: alt.altitudeDeg,
      zone: alt.zone,
      zoneFraction: alt.zoneFraction,
      color,
      gnosticTime: formatGnosticTime(pLo),
    });
  }

  // Group into contiguous bands
  const bands = groupIntoBands(stops);

  // Compute summary fractions
  const summary = computeSummary(stops);

  return {
    source: 'GnosisDaylightTopology',
    observer,
    startEpochMs,
    endEpochMs,
    spanPLo,
    resolution,
    bands,
    gradientStops: stops,
    summary,
  };
}

function groupIntoBands(stops: readonly DaylightGradientStop[]): DaylightBand[] {
  if (stops.length === 0) return [];

  const firstStop = stops[0]!;
  const bands: DaylightBand[] = [];
  let currentZone = firstStop.zone;
  let bandStops: DaylightGradientStop[] = [firstStop];
  let bandStart = firstStop.position;

  for (let i = 1; i < stops.length; i++) {
    const stop = stops[i]!;
    if (stop.zone !== currentZone) {
      const prevStop = stops[i - 1]!;
      bands.push({
        zone: currentZone,
        startPosition: bandStart,
        endPosition: prevStop.position,
        durationMs: prevStop.epochMs - (bandStops[0]?.epochMs ?? prevStop.epochMs),
        gradientStops: bandStops,
      });
      currentZone = stop.zone;
      bandStops = [stop];
      bandStart = stop.position;
    } else {
      bandStops.push(stop);
    }
  }

  // Final band
  const last = stops[stops.length - 1]!;
  const firstOfLastBand = bandStops[0]!;
  bands.push({
    zone: currentZone,
    startPosition: bandStart,
    endPosition: last.position,
    durationMs: last.epochMs - firstOfLastBand.epochMs,
    gradientStops: bandStops,
  });

  return bands;
}

function computeSummary(stops: readonly DaylightGradientStop[]): DaylightSummary {
  if (stops.length === 0) {
    return {
      daylightFraction: 0,
      nightFraction: 0,
      civilTwilightFraction: 0,
      nauticalTwilightFraction: 0,
      astronomicalTwilightFraction: 0,
    };
  }

  const counts: Record<TwilightZone, number> = {
    day: 0,
    civil: 0,
    nautical: 0,
    astronomical: 0,
    night: 0,
  };

  for (const stop of stops) {
    counts[stop.zone]++;
  }

  const total = stops.length;
  return {
    daylightFraction: counts.day / total,
    nightFraction: counts.night / total,
    civilTwilightFraction: counts.civil / total,
    nauticalTwilightFraction: counts.nautical / total,
    astronomicalTwilightFraction: counts.astronomical / total,
  };
}

// ─── CSS / Canvas Output ────────────────────────────────────────────

/**
 * Emit a CSS `linear-gradient()` string from a daylight topology packet.
 * Gradient runs left-to-right (time axis).
 */
export function daylightGradientCSS(packet: DaylightTopologyPacket): string {
  const colorStops = packet.gradientStops.map(
    (stop) => `${colorToRgba(stop.color)} ${(stop.position * 100).toFixed(1)}%`
  );
  return `linear-gradient(to right, ${colorStops.join(', ')})`;
}

/**
 * Emit gradient stops for Canvas2D `createLinearGradient().addColorStop()`.
 */
export function daylightGradientStopsForCanvas(
  packet: DaylightTopologyPacket
): Array<{ offset: number; color: string }> {
  return packet.gradientStops.map((stop) => ({
    offset: stop.position,
    color: colorToRgba(stop.color),
  }));
}

// ─── Spacetime Surface ──────────────────────────────────────────────

export interface CreateSpacetimeSurfaceInput {
  readonly latitudeDeg: number;
  readonly startEpochMs?: number;
  readonly spanPLo?: number;
  readonly longitudeBins?: number;
  readonly timeBins?: number;
  readonly palette?: DaylightPalette;
}

/**
 * Create a 2D spacetime surface: longitude (x) vs time (y).
 *
 * Shows the daylight terminator sweeping across Earth as time advances.
 * Diagonal bands of day/night visualize how spacetime relates to daylight.
 *
 * Defaults:
 *   - longitudeBins: 55 (Pleroma -- ~6.5 degree bins)
 *   - timeBins: 55 (Pleroma -- one row per ~82 minutes for 1 pLo span)
 */
export function createSpacetimeSurface(
  input: CreateSpacetimeSurfaceInput
): SpacetimeSurface {
  const {
    latitudeDeg,
    startEpochMs = Date.now(),
    spanPLo = 1,
    longitudeBins = GNOSTIC.PLEROMA,
    timeBins = GNOSTIC.PLEROMA,
    palette = DEFAULT_DAYLIGHT_PALETTE,
  } = input;

  const totalMs = spanPLo * PICOLORENZO_MS;
  const endEpochMs = startEpochMs + totalMs;
  const timeStep = totalMs / Math.max(1, timeBins - 1);
  const lonStep = 360 / longitudeBins;

  const cells: SpacetimeCell[] = [];

  for (let t = 0; t < timeBins; t++) {
    const epochMs = startEpochMs + t * timeStep;
    for (let l = 0; l < longitudeBins; l++) {
      const longitudeDeg = -180 + l * lonStep;
      const observer: SolarObserver = { latitudeDeg, longitudeDeg };
      const alt = solarPosition(observer, epochMs);
      const color = interpolateDaylightColor(palette, alt.altitudeDeg);

      cells.push({
        longitudeDeg,
        epochMs,
        altitudeDeg: alt.altitudeDeg,
        zone: alt.zone,
        color,
      });
    }
  }

  return {
    source: 'GnosisDaylightSpacetime',
    longitudeBins,
    timeBins,
    startEpochMs,
    endEpochMs,
    latitudeDeg,
    cells,
  };
}

// ─── Daylight Mesh (quantized, aeon-weather compatible) ─────────────

export interface CreateDaylightMeshInput {
  readonly epochMs?: number;
  readonly resolutionDeg?: number;
  readonly palette?: DaylightPalette;
  /** Optional lat bounds to limit computation. Default: full globe. */
  readonly minLat?: number;
  readonly maxLat?: number;
  /** Optional lon bounds. Default: full globe. */
  readonly minLon?: number;
  readonly maxLon?: number;
}

/**
 * Create a quantized daylight mesh aligned to the aeon-weather mesh grid.
 *
 * Default resolution: 0.25 degrees (matching storms-watch mesh).
 * For global coverage at 0.25 deg, this produces 721 x 1440 = 1,038,240 cells.
 * Use lat/lon bounds to limit the computation to a region of interest.
 *
 * Each cell carries `isDaylight`, the twilight zone, altitude, and color.
 * The mesh can be joined with aeon-weather MeshCoordinate by latIndex/lonIndex.
 */
export function createDaylightMesh(
  input: CreateDaylightMeshInput = {}
): DaylightMeshPacket {
  const {
    epochMs = Date.now(),
    resolutionDeg = 0.25,
    palette = DEFAULT_DAYLIGHT_PALETTE,
    minLat = -90,
    maxLat = 90,
    minLon = -180,
    maxLon = 180,
  } = input;

  const latCount = Math.round((maxLat - minLat) / resolutionDeg) + 1;
  const lonCount = Math.round((maxLon - minLon) / resolutionDeg);

  const cells: DaylightMeshCell[] = [];
  let daylightCount = 0;

  for (let li = 0; li < latCount; li++) {
    const lat = minLat + li * resolutionDeg;
    const latIdx = Math.round((lat + 90) / resolutionDeg);

    for (let lo = 0; lo < lonCount; lo++) {
      const lon = minLon + lo * resolutionDeg;
      const lonIdx = Math.round(((lon + 180) % 360) / resolutionDeg);

      const observer: SolarObserver = { latitudeDeg: lat, longitudeDeg: lon };
      const alt = solarPosition(observer, epochMs);
      const color = interpolateDaylightColor(palette, alt.altitudeDeg);
      const daylit = alt.zone === 'day';

      if (daylit) daylightCount++;

      cells.push({
        latIndex: latIdx,
        lonIndex: lonIdx,
        centroidLat: lat,
        centroidLon: lon,
        altitudeDeg: alt.altitudeDeg,
        zone: alt.zone,
        isDaylight: daylit,
        color,
      });
    }
  }

  const totalCells = latCount * lonCount;

  return {
    source: 'GnosisDaylightMesh',
    epochMs,
    resolutionDeg,
    latCount,
    lonCount,
    cells,
    daylightFraction: totalCells > 0 ? daylightCount / totalCells : 0,
  };
}
