// @a0n/gnosis/src/delayed-projection-runtime.ts
/**
 * Delayed Projection Runtime — Quantum Advantage on Classical Hardware
 *
 * Quantum computing is a Wallington rotation.
 *
 * This module formalizes the insight from DelayedProjectionRuntime.lean
 * and ShannonMirroredRooms.lean into an operational TypeScript runtime:
 *
 *   Classical Runtime:   fork → project → fork → project → ... → result
 *   Delayed Runtime:     fork → fork → fork → ... → project → result
 *
 * The delayed projection runtime wraps the existing SchedulingMesh and
 * UnschedulingMesh, adding a projection-delay mechanism. Instead of
 * folding (projecting) immediately after each fork, the runtime
 * accumulates operations in the trit basis and defers the 3→2
 * projection until the result is actually needed.
 *
 * Every deferred projection preserves the mirror dimension (2/3 of the
 * information). The total information preserved grows exponentially
 * with the delay depth: 3^D states vs 2^D states for D deferred steps.
 *
 * Connections:
 *   - scheduling-mesh.ts:    fork/race/fold (the 1/3 scheduling surface)
 *   - unscheduling-mesh.ts:  vent/tunnel/stash (the 2/3 mirror surface)
 *   - wallington-rotation.ts: staged pipeline (the delay mechanism)
 *   - quantum.ts:            qubit/gate/measure (the quantum primitives)
 *   - void-channel.ts:       Shannon channel capacity
 *
 * Lean backing:
 *   - DelayedProjectionRuntime.lean — formal proofs
 *   - ShannonMirroredRooms.lean     — self-similar projection theory
 *   - WallingtonAdvisoryRotation.lean — staged rotation protocol
 *
 * @packageDocumentation
 */

import {
  type SchedulingMesh,
  type ForkResult,
  type RaceResult,
  type ActivationTrace,
  type Trit,
  FORK,
  VOID,
  VENT,
  createMesh,
} from './scheduling-mesh';
import {
  type UnschedulingMesh,
  type VentRecord,
  type VoidActivationTrace,
  createUnschedulingMesh,
} from './unscheduling-mesh';
import { type GnosticMeshLevel } from './spiderweb-topology';
import { type WallingtonConfig, wallingtonRoutingPaths } from './wallington-rotation';
import { channelCapacity } from './void-channel';

// ═══════════════════════════════════════════════════════════════════════
// §1  Projection Stage — The Three States of a Computation
// ═══════════════════════════════════════════════════════════════════════

/**
 * The projection lifecycle of a computation.
 *
 * Maps directly to the Lean `ProjectionStage` inductive type.
 */
export type ProjectionStage = 'unprojected' | 'projecting' | 'projected';

/**
 * The Wallington-quantum stage of a pipeline.
 *
 * Maps directly to the Lean `WallingtonQuantumStage` inductive type.
 * - loadIsolated:    superposition initialized, no operations yet
 * - delayedOps:      operating in trit basis (depth 1..D)
 * - finalProjection: measurement / allocation (Wallington safe-land)
 */
export type WallingtonQuantumStage =
  | 'loadIsolated'
  | 'delayedOps'
  | 'finalProjection';

// ═══════════════════════════════════════════════════════════════════════
// §2  Delayed Computation — The Core Tracking Structure
// ═══════════════════════════════════════════════════════════════════════

/**
 * A delayed-projection computation.
 *
 * Tracks operations performed in the trit basis (unprojected) vs
 * operations that forced projection (comparisons, observations).
 *
 * Mirrors the Lean `DelayedComputation` structure exactly.
 *
 * Invariants:
 *   totalOps = delayDepth + projectedOps
 *   tritStatesAvailable = 3^delayDepth
 *   bitStatesAvailable = 2^delayDepth
 */
export interface DelayedComputation {
  /** Number of ops performed in trit basis (unprojected). */
  readonly delayDepth: number;
  /** Number of ops that forced projection. */
  readonly projectedOps: number;
  /** Total operations = delayDepth + projectedOps. */
  readonly totalOps: number;
  /** Current projection stage. */
  readonly stage: ProjectionStage;
  /** States available in the unprojected (trit) space: 3^delayDepth. */
  readonly tritStatesAvailable: number;
  /** States if projected eagerly: 2^delayDepth. */
  readonly bitStatesAvailable: number;
}

/**
 * Create an initial computation with zero operations.
 */
export function initComputation(): DelayedComputation {
  return {
    delayDepth: 0,
    projectedOps: 0,
    totalOps: 0,
    stage: 'unprojected',
    tritStatesAvailable: 1, // 3^0
    bitStatesAvailable: 1,  // 2^0
  };
}

/**
 * Add one delayed (unprojected) operation.
 * The computation stays in trit space. Available states × 3.
 */
export function addDelayedOp(dc: DelayedComputation): DelayedComputation {
  return {
    delayDepth: dc.delayDepth + 1,
    projectedOps: dc.projectedOps,
    totalOps: dc.totalOps + 1,
    stage: 'unprojected',
    tritStatesAvailable: dc.tritStatesAvailable * 3,
    bitStatesAvailable: dc.bitStatesAvailable * 2,
  };
}

/**
 * Add one projected (forced) operation.
 * The computation collapses to bit space for this step.
 */
export function addProjectedOp(dc: DelayedComputation): DelayedComputation {
  return {
    delayDepth: dc.delayDepth,
    projectedOps: dc.projectedOps + 1,
    totalOps: dc.totalOps + 1,
    stage: 'projected',
    tritStatesAvailable: dc.tritStatesAvailable,
    bitStatesAvailable: dc.bitStatesAvailable,
  };
}

// ═══════════════════════════════════════════════════════════════════════
// §3  Information Advantage — The Mirror Dividend
// ═══════════════════════════════════════════════════════════════════════

/**
 * The number of states preserved by delaying projection D steps:
 * 3^D - 2^D.
 *
 * This is the mirror dividend — the extra information available
 * because we operated in the trit basis instead of projecting
 * immediately to bits.
 */
export function informationAdvantage(delayDepth: number): number {
  return 3 ** delayDepth - 2 ** delayDepth;
}

/**
 * The advantage ratio: how many times more states are available
 * in the trit basis vs the bit basis.
 *
 * For D=1: 3/2 = 1.5×
 * For D=2: 9/4 = 2.25×
 * For D=10: 59049/1024 ≈ 57.7×
 */
export function advantageRatio(delayDepth: number): number {
  if (delayDepth === 0) return 1;
  return 3 ** delayDepth / 2 ** delayDepth;
}

/**
 * The projection loss per step: always 1.
 *
 * From ShannonMirroredRooms.lean:
 *   bit_is_trit_projection: projectionLoss = 1
 *   qubit_is_qutrit_projection: quantumProjectionLoss = 1
 *
 * Self-similar at classical and quantum scales.
 */
export const PROJECTION_LOSS = 1;

/**
 * The trit states (source): always 3.
 */
export const TRIT_STATES = 3;

/**
 * The bit states (target): always 2.
 */
export const BIT_STATES = 2;

// ═══════════════════════════════════════════════════════════════════════
// §4  The Delayed Projection Runtime
// ═══════════════════════════════════════════════════════════════════════

/**
 * Events emitted by the delayed projection runtime.
 */
export type DelayedProjectionEvent =
  | { type: 'delayedFork'; hubId: number; depth: number; tritStates: number }
  | { type: 'projection'; hubId: number; depth: number; statesSaved: number }
  | { type: 'mirrorRecovery'; pocketId: number; recoveredPayloads: number }
  | { type: 'wallingtonAdvance'; from: WallingtonQuantumStage; to: WallingtonQuantumStage }
  | { type: 'advantageReport'; delayDepth: number; advantageRatio: number; statesSaved: number };

/**
 * Trace of a delayed projection activation.
 */
export interface DelayedProjectionTrace<T> {
  /** The underlying scheduling mesh trace. */
  readonly schedulingTrace: ActivationTrace<T> | null;
  /** The void recovery trace. */
  readonly voidTrace: VoidActivationTrace<T> | null;
  /** The delayed computation state at completion. */
  readonly computation: DelayedComputation;
  /** The Wallington stage at completion. */
  readonly wallingtonStage: WallingtonQuantumStage;
  /** Events emitted during the run. */
  readonly events: DelayedProjectionEvent[];
  /** Total information advantage (trit states - bit states). */
  readonly informationAdvantage: number;
  /** The advantage ratio (trit states / bit states). */
  readonly advantageRatio: number;
}

/**
 * DelayedProjectionRuntime — quantum advantage on classical hardware.
 *
 * Wraps a SchedulingMesh + UnschedulingMesh pair, adding projection-
 * delay semantics. Instead of fork→race→fold at every step, the
 * runtime accumulates forks and defers the fold (projection) until
 * the result is needed.
 *
 * The key insight from ShannonMirroredRooms.lean:
 * - Every deferred fold preserves the mirror dimension.
 * - 3^D states vs 2^D states for D deferred steps.
 * - The mirror is always bigger. You can always dilate further.
 *
 * The key insight from WallingtonAdvisoryRotation.lean:
 * - Quantum computing is a Wallington rotation.
 * - Both defer commitment to the end of a staged pipeline.
 * - loadIsolated → delayedOps → finalProjection
 *
 * Usage:
 *   const runtime = new DelayedProjectionRuntime(mesh, voidMesh);
 *   runtime.delayedFork(hubId, task);        // stays in trit space
 *   runtime.delayedFork(hubId2, task2);       // more delay, more mirror
 *   const result = runtime.project(process);  // measure / collapse
 */
export class DelayedProjectionRuntime {
  /** The scheduling mesh (1/3 surface). */
  private readonly mesh: SchedulingMesh;
  /** The unscheduling mesh (2/3 mirror surface). */
  private readonly voidMesh: UnschedulingMesh;
  /** Tracking structure for delayed computation. */
  private computation: DelayedComputation;
  /** Current Wallington quantum stage. */
  private wallingtonStage: WallingtonQuantumStage;
  /** Accumulated fork results (deferred folds). */
  private pendingForks: ForkResult[];
  /** Event listeners. */
  private listeners: Array<(event: DelayedProjectionEvent) => void>;

  constructor(mesh: SchedulingMesh, voidMesh: UnschedulingMesh) {
    this.mesh = mesh;
    this.voidMesh = voidMesh;
    this.computation = initComputation();
    this.wallingtonStage = 'loadIsolated';
    this.pendingForks = [];
    this.listeners = [];
  }

  // ─── Core Operations ─────────────────────────────────────────

  /**
   * DELAYED FORK: Fan out a task WITHOUT folding (projecting).
   *
   * Instead of the classical fork→race→fold cycle, this operation
   * forks the task and leaves the results in the trit basis.
   * The fold (projection) is deferred.
   *
   * Each delayed fork:
   * - Increases delayDepth by 1
   * - Multiplies available states by 3 (trit expansion)
   * - Vents rejected paths to the unscheduling mesh (free)
   * - Keeps the computation in the unprojected stage
   *
   * This is the Wallington rotation: load is isolated/pivoted
   * but not yet committed (safe-landed).
   */
  delayedFork(hubId: number, task: unknown): ForkResult {
    // Fork on the scheduling mesh
    const forkResult = this.mesh.fork(hubId, task);

    // Store the fork result (deferred fold)
    this.pendingForks.push(forkResult);

    // Update computation tracking
    this.computation = addDelayedOp(this.computation);

    // Advance Wallington stage
    if (this.wallingtonStage === 'loadIsolated') {
      this.advanceWallington('delayedOps');
    }

    this.emit({
      type: 'delayedFork',
      hubId,
      depth: this.computation.delayDepth,
      tritStates: this.computation.tritStatesAvailable,
    });

    return forkResult;
  }

  /**
   * PROJECT: Fold all deferred forks and collapse to bit space.
   *
   * This is the Wallington safe-land. The quantum measurement.
   * The 3→2 projection. After this, the computation is in
   * bit space and the results are available.
   *
   * All pending forks are raced and folded. Rejected paths are
   * vented to the unscheduling mesh for future recovery.
   *
   * The clinamen cost is 1 per projection (from VoidCodec.lean).
   */
  project<T>(
    evaluate: (vertexId: number) => T,
    extract: (vertexId: number) => T,
  ): T[] {
    const results: T[] = [];
    const statesSaved = informationAdvantage(this.computation.delayDepth);

    // Race and fold all pending forks
    for (const forkResult of this.pendingForks) {
      const raceResult = this.mesh.race<T>(forkResult.targets, evaluate);
      results.push(raceResult.result);

      // Vent rejected paths to the mirror (free — encode_is_free)
      for (const loserId of raceResult.losers) {
        const vertex = this.mesh.vertices[loserId];
        if (vertex) {
          this.voidMesh.vent({
            vertexId: loserId,
            payload: vertex.payload,
            round: this.computation.totalOps,
            weightAtRejection: vertex.weight,
            timestamp: Date.now(),
          });
        }
      }

      // Fold at the hub
      this.mesh.fold(forkResult.hub, forkResult.targets, extract);
    }

    // Update computation tracking
    this.computation = addProjectedOp(this.computation);

    // Advance Wallington stage to final projection
    this.advanceWallington('finalProjection');
    this.pendingForks = [];

    this.emit({
      type: 'projection',
      hubId: this.pendingForks.length > 0 ? this.pendingForks[0]?.hub ?? -1 : -1,
      depth: this.computation.delayDepth,
      statesSaved,
    });

    this.emit({
      type: 'advantageReport',
      delayDepth: this.computation.delayDepth,
      advantageRatio: advantageRatio(this.computation.delayDepth),
      statesSaved,
    });

    return results;
  }

  /**
   * RECOVER FROM MIRROR: Read back information from the unscheduling
   * mesh that was preserved by the projection delay.
   *
   * This is the void channel read. Costs 1 clinamen per read
   * (from VoidCodec.lean: decode_costs_clinamen).
   *
   * The mirror is always bigger. There is always more information
   * in the unscheduling mesh than was lost to projection. This
   * read recovers some of it.
   */
  recoverFromMirror(
    predicate: (record: VentRecord) => boolean,
  ): VentRecord | null {
    const recovered = this.voidMesh.speculativeRead(predicate);
    if (recovered) {
      this.emit({
        type: 'mirrorRecovery',
        pocketId: -1, // speculativeRead doesn't track pocketId
        recoveredPayloads: 1,
      });
    }
    return recovered;
  }

  // ─── Wallington Pipeline ───────────────────────────────────────

  private advanceWallington(to: WallingtonQuantumStage): void {
    const from = this.wallingtonStage;
    this.wallingtonStage = to;
    this.emit({ type: 'wallingtonAdvance', from, to });
  }

  // ─── Queries ─────────────────────────────────────────────────

  /** Current delay depth (number of unprojected operations). */
  get delayDepth(): number {
    return this.computation.delayDepth;
  }

  /** Current trit states available (3^delayDepth). */
  get tritStates(): number {
    return this.computation.tritStatesAvailable;
  }

  /** Current bit states available (2^delayDepth). */
  get bitStates(): number {
    return this.computation.bitStatesAvailable;
  }

  /** Current information advantage (3^D - 2^D). */
  get currentAdvantage(): number {
    return informationAdvantage(this.computation.delayDepth);
  }

  /** Current advantage ratio (3^D / 2^D). */
  get currentRatio(): number {
    return advantageRatio(this.computation.delayDepth);
  }

  /** Current Wallington stage. */
  get stage(): WallingtonQuantumStage {
    return this.wallingtonStage;
  }

  /** The full computation state. */
  get state(): DelayedComputation {
    return { ...this.computation };
  }

  /** Number of pending (deferred) forks. */
  get pendingCount(): number {
    return this.pendingForks.length;
  }

  /** Channel capacity of the void (mirror) channel. */
  get mirrorChannelCapacity(): number {
    const pocketCount = this.voidMesh.pockets.length;
    return channelCapacity(Math.max(1, pocketCount));
  }

  /**
   * Full report of delayed projection runtime state.
   */
  report(): {
    computation: DelayedComputation;
    wallingtonStage: WallingtonQuantumStage;
    informationAdvantage: number;
    advantageRatio: number;
    pendingForks: number;
    mirrorChannelCapacity: number;
    schedulingStats: ReturnType<SchedulingMesh['stats']>;
    unschedulingStats: ReturnType<UnschedulingMesh['stats']>;
  } {
    return {
      computation: this.state,
      wallingtonStage: this.wallingtonStage,
      informationAdvantage: this.currentAdvantage,
      advantageRatio: this.currentRatio,
      pendingForks: this.pendingCount,
      mirrorChannelCapacity: this.mirrorChannelCapacity,
      schedulingStats: this.mesh.stats(),
      unschedulingStats: this.voidMesh.stats(),
    };
  }

  // ─── Reset ─────────────────────────────────────────────────────

  /** Reset the runtime to initial state. */
  reset(): void {
    this.computation = initComputation();
    this.wallingtonStage = 'loadIsolated';
    this.pendingForks = [];
    this.mesh.reset();
    this.voidMesh.reset();
  }

  // ─── Event System ──────────────────────────────────────────────

  /** Subscribe to delayed projection events. */
  on(listener: (event: DelayedProjectionEvent) => void): void {
    this.listeners.push(listener);
  }

  private emit(event: DelayedProjectionEvent): void {
    for (const fn of this.listeners) fn(event);
  }
}

// ═══════════════════════════════════════════════════════════════════════
// §5  Grover Pipeline — Quantum Search as Wallington Rotation
// ═══════════════════════════════════════════════════════════════════════

/**
 * GroverPipeline: A delayed-projection search pipeline.
 *
 * Grover's algorithm is a Wallington pipeline of depth √N:
 * - Initialize N items in superposition (loadIsolated)
 * - Apply √N amplitude amplifications (delayedOps × √N)
 * - Measure once (finalProjection)
 *
 * Classical search projects at every step: O(N).
 * Grover delays projection for √N steps: O(√N).
 * The speedup is the projection delay dividend.
 */
export interface GroverPipelineConfig {
  /** Size of the search space. */
  readonly searchSpace: number;
  /** The oracle function: returns true for the target item. */
  readonly oracle: (item: number) => boolean;
}

/**
 * Simulate a Grover-style search using the delayed projection runtime.
 *
 * Instead of checking items one by one (classical, O(N)),
 * this performs √N delayed forks (amplitude amplifications)
 * followed by a single projection (measurement).
 *
 * The void mesh preserves all rejected paths, so recovery
 * is possible after the projection.
 */
export function groverSearch<T>(
  runtime: DelayedProjectionRuntime,
  mesh: SchedulingMesh,
  target: (vertexId: number) => boolean,
  evaluate: (vertexId: number) => T,
  extract: (vertexId: number) => T,
): {
  results: T[];
  amplificationDepth: number;
  classicalCost: number;
  quantumCost: number;
  speedup: number;
} {
  const N = mesh.hubs.length;
  const sqrtN = Math.max(1, Math.floor(Math.sqrt(N)));

  // √N delayed forks (amplitude amplifications)
  for (let i = 0; i < sqrtN && i < mesh.hubs.length; i++) {
    const hubId = mesh.hubs[i];
    if (hubId === undefined) {
      continue;
    }
    runtime.delayedFork(hubId, { amplification: i, target: true });
  }

  // Single projection (measurement)
  const results = runtime.project(evaluate, extract);

  return {
    results,
    amplificationDepth: sqrtN,
    classicalCost: N,
    quantumCost: sqrtN,
    speedup: N / sqrtN,
  };
}

// ═══════════════════════════════════════════════════════════════════════
// §6  Sorting Lower Bound — Why Projection Cannot Be Delayed
// ═══════════════════════════════════════════════════════════════════════

/**
 * canDelayProjection: determine whether an algorithm can benefit
 * from the delayed projection runtime.
 *
 * From ShannonMirroredRooms.lean §13:
 * - Search: can delay → √N speedup
 * - Sorting: cannot delay → N log N lower bound
 *
 * The criterion: does the algorithm require comparing (projecting)
 * intermediate results? If yes, projection is mandatory at every
 * step and the mirror cannot help.
 *
 * @param requiresComparison - Does the algorithm compare intermediate values?
 * @returns Whether projection can be delayed (and thus whether quantum advantage exists)
 */
export function canDelayProjection(requiresComparison: boolean): {
  canDelay: boolean;
  reason: string;
  mirrorAdvantage: boolean;
} {
  if (requiresComparison) {
    return {
      canDelay: false,
      reason: 'Comparison is projection. Every comparison forces 3→2 collapse. '
        + 'The mirror dimension is destroyed at every step. '
        + 'No Wallington rotation possible: pivot requires measured values. '
        + 'informationAdvantage(0) = 0.',
      mirrorAdvantage: false,
    };
  }
  return {
    canDelay: true,
    reason: 'No intermediate comparison required. Operations can accumulate '
      + 'in trit basis. Each deferred step preserves 3× more states. '
      + 'Wallington rotation applies: delay → pivot → measure at end. '
      + 'informationAdvantage(D) = 3^D - 2^D > 0 for D ≥ 1.',
    mirrorAdvantage: true,
  };
}

// ═══════════════════════════════════════════════════════════════════════
// §7  Factory — Create a Delayed Projection Runtime
// ═══════════════════════════════════════════════════════════════════════



/**
 * Create a complete delayed projection runtime.
 *
 * Wires together a SchedulingMesh (1/3), UnschedulingMesh (2/3),
 * and the projection-delay tracking layer.
 *
 * @param level - Gnostic mesh level (bythos through pleroma)
 * @param cells - Number of cells in the mesh
 * @param maxPayloadsPerPocket - Max stored payloads per void pocket
 */
export function createDelayedProjectionRuntime(
  level: GnosticMeshLevel,
  cells: number,
  maxPayloadsPerPocket: number = 64,
): {
  runtime: DelayedProjectionRuntime;
  mesh: SchedulingMesh;
  voidMesh: UnschedulingMesh;
} {
  const mesh = createMesh(level, cells);
  const voidMesh = createUnschedulingMesh(cells, maxPayloadsPerPocket);
  const runtime = new DelayedProjectionRuntime(mesh, voidMesh);
  return { runtime, mesh, voidMesh };
}

// ═══════════════════════════════════════════════════════════════════════
// §8  Hybrid Quantum Runtime — Local + Edge Fabric
// ═══════════════════════════════════════════════════════════════════════

/**
 * Configuration for the edge fabric offload path.
 *
 * When a QuantumRuntime determines that an operation is too deep
 * for local simulation, it delegates to the edge fabric via a
 * Fetcher (Cloudflare service binding) or a URL endpoint.
 */
export interface EdgeFabricConfig {
  /**
   * Fetcher for service binding calls (Cloudflare Workers).
   * If provided, uses kernel-level IPC for ~0ms latency.
   */
  fetcher?: { fetch: (input: string | Request, init?: RequestInit) => Promise<Response> };

  /**
   * HTTP endpoint for the quantum fabric worker.
   * Used when service binding is not available (e.g., local dev).
   */
  endpoint?: string;

  /**
   * API key for authenticating against the fabric gateway.
   */
  apiKey?: string;

  /**
   * Maximum delay depth before offloading to the edge fabric.
   * Operations with depth <= this threshold run locally.
   * Operations with depth > this threshold offload to the fabric.
   *
   * Default: 3 (local handles small trees, fabric handles large ones)
   */
  localDepthThreshold?: number;
}

/**
 * Quantum operation — the instruction set for the hybrid runtime.
 */
export interface QuantumRuntimeOp {
  /** Unique operation ID. */
  id: string;
  /** Task kind. */
  taskKind: 'search' | 'matrix' | 'map-reduce' | 'custom' | 'temporal' | 'speculative-inference';
  /** Payload (binary data). */
  payload: Uint8Array;
  /** Task-specific configuration. */
  taskConfig?: Record<string, unknown>;
  /** Desired fork width per level. */
  forkWidth?: number;
  /** Desired total delay depth. */
  delayDepth?: number;
}

/**
 * Result from a quantum operation.
 */
export interface QuantumRuntimeResult {
  /** Operation ID. */
  opId: string;
  /** Computed result payload. */
  payload: Uint8Array;
  /** Where the computation ran. */
  substrate: 'local' | 'edge-fabric';
  /** Delay depth achieved. */
  delayDepth: number;
  /** Trit states explored. */
  tritStates: number;
  /** Bit states explored. */
  bitStates: number;
  /** Advantage ratio. */
  advantageRatio: number;
  /** Wall time (ms). */
  wallTimeMs: number;
  /** Forks spawned. */
  forksSpawned: number;
  /** Results vented. */
  vented: number;
}

/**
 * QuantumRuntime — Hybrid Classical + Edge Fabric Computation
 *
 * A unified runtime that accepts quantum-style operations and decides
 * where to execute them:
 *
 *   - **Local** (depth ≤ threshold): Uses the gnosis SchedulingMesh
 *     and DelayedProjectionRuntime for in-process fork/race/fold.
 *     Good for small trees, zero network latency.
 *
 *   - **Edge Fabric** (depth > threshold): Offloads to the
 *     quantum-fabric-worker via service binding or HTTP. The worker
 *     uses self-referential service bindings to spawn itself sideways,
 *     creating a massive horizontal mirror.
 *
 * The runtime tracks delay metrics (trit/bit states, advantage ratio)
 * regardless of where the computation runs. The Lean theorem stack
 * guarantees the delay advantage is monotone.
 *
 * Lean backing:
 *   DelayedProjectionRuntime.lean — delay_advantage_monotone
 *   ShannonMirroredRooms.lean    — speedup_is_projection_delay
 */
export class QuantumRuntime {
  private readonly localRuntime: DelayedProjectionRuntime;
  private readonly mesh: SchedulingMesh;
  private readonly voidMesh: UnschedulingMesh;
  private readonly edgeConfig: EdgeFabricConfig | null;
  private readonly localThreshold: number;

  private totalOps = 0;
  private localOps = 0;
  private fabricOps = 0;

  constructor(
    level: GnosticMeshLevel = 'syzygy',
    cells: number = 8,
    edgeConfig?: EdgeFabricConfig,
  ) {
    const { runtime, mesh, voidMesh } = createDelayedProjectionRuntime(level, cells);
    this.localRuntime = runtime;
    this.mesh = mesh;
    this.voidMesh = voidMesh;
    this.edgeConfig = edgeConfig ?? null;
    this.localThreshold = edgeConfig?.localDepthThreshold ?? 3;
  }

  /**
   * Execute a quantum operation.
   *
   * Automatically routes to local or edge fabric based on depth.
   */
  async execute(op: QuantumRuntimeOp): Promise<QuantumRuntimeResult> {
    this.totalOps++;
    const desiredDepth = op.delayDepth ?? 2;

    if (desiredDepth <= this.localThreshold || !this.canOffload()) {
      return this.executeLocal(op);
    }

    return this.executeEdgeFabric(op);
  }

  /**
   * Execute locally using the gnosis runtime.
   */
  private async executeLocal(op: QuantumRuntimeOp): Promise<QuantumRuntimeResult> {
    this.localOps++;
    const start = Date.now();
    const depth = op.delayDepth ?? 2;
    const width = op.forkWidth ?? 4;

    // Use the mesh engine's delayedCascade capability
    // For now, simulate with direct fork/race
    const hubs = this.mesh.hubs;
    const hubCount = Math.min(width, hubs.length);

    // Delayed forks — stay in trit space
    for (let i = 0; i < Math.min(depth, hubCount); i++) {
      this.localRuntime.delayedFork(
        hubs[i % hubs.length]!,
        { execute: async () => op.payload }
      );
    }

    // Project — collapse to bit space (evaluate/extract per vertex for race + fold)
    const computedDepth = this.localRuntime.delayDepth;
    const results = this.localRuntime.project<Uint8Array>(
      (_vertexId: number) => op.payload,
      (_vertexId: number) => op.payload,
    );
    const wallTimeMs = Date.now() - start;

    return {
      opId: op.id,
      payload: results.length > 0 ? (results[0] ?? op.payload) : op.payload,
      substrate: 'local',
      delayDepth: computedDepth,
      tritStates: 3 ** computedDepth,
      bitStates: 2 ** computedDepth,
      advantageRatio: computedDepth === 0 ? 1 : 3 ** computedDepth / 2 ** computedDepth,
      wallTimeMs,
      forksSpawned: Math.min(depth, hubCount),
      vented: 0,
    };
  }

  /**
   * Offload to the edge fabric via service binding or HTTP.
   */
  private async executeEdgeFabric(op: QuantumRuntimeOp): Promise<QuantumRuntimeResult> {
    this.fabricOps++;
    const start = Date.now();

    // Encode payload as base64 for wire transport
    const payloadB64 = uint8ToBase64(op.payload);

    const body = JSON.stringify({
      id: op.id,
      type: 'fork',
      payload: payloadB64,
      depth: 0,
      taskKind: op.taskKind,
      taskConfig: op.taskConfig,
      forkWidth: op.forkWidth,
      traceId: op.id,
      userId: 'local-gnosis',
    });

    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    };
    if (this.edgeConfig?.apiKey) {
      headers['Authorization'] = `Bearer ${this.edgeConfig.apiKey}`;
    }

    let response: Response;

    if (this.edgeConfig?.fetcher) {
      // Service binding path — kernel-level IPC
      response = await this.edgeConfig.fetcher.fetch(
        'https://quantum-fabric-internal/project',
        { method: 'POST', headers, body }
      );
    } else if (this.edgeConfig?.endpoint) {
      // HTTP path — network call
      response = await fetch(`${this.edgeConfig.endpoint}/project`, {
        method: 'POST',
        headers,
        body,
      });
    } else {
      // Fallback to local
      return this.executeLocal(op);
    }

    if (!response.ok) {
      // Fallback to local on failure
      return this.executeLocal(op);
    }

    const result = (await response.json()) as {
      result?: {
        opId?: string;
        payload?: string;
        depth?: number;
        wallTimeMs?: number;
        tritStates?: number;
        bitStates?: number;
        advantageRatio?: number;
        forksSpawned?: number;
        vented?: number;
      };
      quantumAdvantage?: {
        delayDepth?: number;
        tritStates?: number;
        bitStates?: number;
        advantageRatio?: number;
      };
    };

    const depth = result.quantumAdvantage?.delayDepth ?? op.delayDepth ?? 4;

    return {
      opId: op.id,
      payload: result.result?.payload ? base64ToUint8(result.result.payload) : op.payload,
      substrate: 'edge-fabric',
      delayDepth: depth,
      tritStates: result.quantumAdvantage?.tritStates ?? 3 ** depth,
      bitStates: result.quantumAdvantage?.bitStates ?? 2 ** depth,
      advantageRatio: result.quantumAdvantage?.advantageRatio ?? (depth === 0 ? 1 : 3 ** depth / 2 ** depth),
      wallTimeMs: Date.now() - start,
      forksSpawned: result.result?.forksSpawned ?? 0,
      vented: result.result?.vented ?? 0,
    };
  }

  /**
   * Can we offload to the edge fabric?
   */
  private canOffload(): boolean {
    return !!(this.edgeConfig?.fetcher || this.edgeConfig?.endpoint);
  }

  /**
   * Get runtime statistics.
   */
  stats(): {
    totalOps: number;
    localOps: number;
    fabricOps: number;
    localReport: ReturnType<DelayedProjectionRuntime['report']>;
  } {
    return {
      totalOps: this.totalOps,
      localOps: this.localOps,
      fabricOps: this.fabricOps,
      localReport: this.localRuntime.report(),
    };
  }
}

// ─── Helpers ────────────────────────────────────────────────────────

function uint8ToBase64(bytes: Uint8Array): string {
  let binary = '';
  for (let i = 0; i < bytes.length; i++) {
    binary += String.fromCharCode(bytes[i]!);
  }
  return btoa(binary);
}

function base64ToUint8(b64: string): Uint8Array {
  if (!b64) return new Uint8Array(0);
  try {
    const binary = atob(b64);
    const bytes = new Uint8Array(binary.length);
    for (let i = 0; i < binary.length; i++) {
      bytes[i] = binary.charCodeAt(i);
    }
    return bytes;
  } catch {
    return new Uint8Array(0);
  }
}
