// @a0n/gnosis/src/differentiable.ts
/**
 * differentiable.ts -- Gradients and parameter updates as topology expressions
 *
 * Differentiable values, parameters, and loss-producing nodes are explicit.
 * Forward and backward passes are inspectable topology transitions.
 *
 * Roadmap item #6: reuses the value/effect/verification model.
 */

// ============================================================================
// Differentiable value -- a value that carries gradient information
// ============================================================================

export interface DiffValue {
  /** The forward value */
  data: number;
  /** Accumulated gradient (null = no gradient) */
  grad: number | null;
  /** Whether this value requires gradient computation */
  requiresGrad: boolean;
  /** Label for inspection */
  label?: string;
}

/**
 * Handles the gnosis diff Value workflow.
 */
export function diffValue(
  data: number,
  requiresGrad: boolean = false,
  label?: string
): DiffValue {
  return { data, grad: requiresGrad ? 0 : null, requiresGrad, label };
}

/**
 * Handles the gnosis parameter workflow.
 */
export function parameter(data: number, label?: string): DiffValue {
  return diffValue(data, true, label);
}

/**
 * Handles the gnosis constant workflow.
 */
export function constant(data: number, label?: string): DiffValue {
  return diffValue(data, false, label);
}

// ============================================================================
// Tape -- reverse-mode autodiff computation graph
// ============================================================================

interface TapeEntry {
  /** Index in the tape */
  index: number;
  /** Indices of parent values */
  parents: number[];
  /** Local gradient function: given output grad, produce parent grads */
  backward: (outGrad: number) => number[];
}

export class GradientTape {
  private values: DiffValue[] = [];
  private tape: TapeEntry[] = [];

  /** Track a leaf value (parameter or input) */
  track(value: DiffValue): number {
    const idx = this.values.length;
    this.values.push(value);
    this.tape.push({ index: idx, parents: [], backward: () => [] });
    return idx;
  }

  /** Record an operation and return the index of the output */
  record(
    result: DiffValue,
    parents: number[],
    backward: (outGrad: number) => number[]
  ): number {
    const idx = this.values.length;
    this.values.push(result);
    this.tape.push({ index: idx, parents, backward });
    return idx;
  }

  /** Get value at index */
  get(idx: number): DiffValue {
    return this.values[idx];
  }

  /** Backward pass: compute gradients from a loss value */
  backward(lossIdx: number): void {
    const grads = new Array(this.values.length).fill(0);
    grads[lossIdx] = 1.0; // dL/dL = 1

    // Reverse topological order
    for (let i = this.tape.length - 1; i >= 0; i--) {
      const entry = this.tape[i];
      if (grads[entry.index] === 0 && i !== lossIdx) continue;

      const parentGrads = entry.backward(grads[entry.index]);
      for (let j = 0; j < entry.parents.length; j++) {
        grads[entry.parents[j]] += parentGrads[j];
      }
    }

    // Store gradients in values
    for (let i = 0; i < this.values.length; i++) {
      if (this.values[i].requiresGrad) {
        this.values[i].grad = grads[i];
      }
    }
  }

  /** Get all parameter values with gradients */
  parameters(): DiffValue[] {
    return this.values.filter((v) => v.requiresGrad);
  }

  /** Reset all gradients to zero */
  zeroGrad(): void {
    for (const v of this.values) {
      if (v.requiresGrad) v.grad = 0;
    }
  }

  /** Number of tracked values */
  get size(): number {
    return this.values.length;
  }
}

// ============================================================================
// Differentiable operations -- forward + backward in one
// ============================================================================

/** Add two values */
export function add(tape: GradientTape, aIdx: number, bIdx: number): number {
  const a = tape.get(aIdx);
  const b = tape.get(bIdx);
  const result = diffValue(a.data + b.data, a.requiresGrad || b.requiresGrad);
  return tape.record(result, [aIdx, bIdx], (g) => [g, g]);
}

/** Multiply two values */
export function mul(tape: GradientTape, aIdx: number, bIdx: number): number {
  const a = tape.get(aIdx);
  const b = tape.get(bIdx);
  const result = diffValue(a.data * b.data, a.requiresGrad || b.requiresGrad);
  return tape.record(result, [aIdx, bIdx], (g) => [g * b.data, g * a.data]);
}

/** Negate a value */
export function neg(tape: GradientTape, aIdx: number): number {
  const a = tape.get(aIdx);
  const result = diffValue(-a.data, a.requiresGrad);
  return tape.record(result, [aIdx], (g) => [-g]);
}

/** Subtract: a - b */
export function sub(tape: GradientTape, aIdx: number, bIdx: number): number {
  return add(tape, aIdx, neg(tape, bIdx));
}

/** ReLU activation */
export function relu(tape: GradientTape, aIdx: number): number {
  const a = tape.get(aIdx);
  const result = diffValue(Math.max(0, a.data), a.requiresGrad);
  return tape.record(result, [aIdx], (g) => [a.data > 0 ? g : 0]);
}

/** Sigmoid activation */
export function sigmoid(tape: GradientTape, aIdx: number): number {
  const a = tape.get(aIdx);
  const s = 1 / (1 + Math.exp(-a.data));
  const result = diffValue(s, a.requiresGrad);
  return tape.record(result, [aIdx], (g) => [g * s * (1 - s)]);
}

/** Tanh activation */
export function tanh(tape: GradientTape, aIdx: number): number {
  const a = tape.get(aIdx);
  const t = Math.tanh(a.data);
  const result = diffValue(t, a.requiresGrad);
  return tape.record(result, [aIdx], (g) => [g * (1 - t * t)]);
}

/** Natural log */
export function log(tape: GradientTape, aIdx: number): number {
  const a = tape.get(aIdx);
  const result = diffValue(Math.log(a.data), a.requiresGrad);
  return tape.record(result, [aIdx], (g) => [g / a.data]);
}

/** Power: a^n */
export function pow(tape: GradientTape, aIdx: number, n: number): number {
  const a = tape.get(aIdx);
  const result = diffValue(a.data ** n, a.requiresGrad);
  return tape.record(result, [aIdx], (g) => [g * n * a.data ** (n - 1)]);
}

/** Exp */
export function exp(tape: GradientTape, aIdx: number): number {
  const a = tape.get(aIdx);
  const e = Math.exp(a.data);
  const result = diffValue(e, a.requiresGrad);
  return tape.record(result, [aIdx], (g) => [g * e]);
}

// ============================================================================
// Loss functions
// ============================================================================

/** Mean squared error loss: (prediction - target)^2 */
export function mseLoss(
  tape: GradientTape,
  predIdx: number,
  targetIdx: number
): number {
  const diff = sub(tape, predIdx, targetIdx);
  return pow(tape, diff, 2);
}

/** Binary cross-entropy loss */
export function bceLoss(
  tape: GradientTape,
  predIdx: number,
  targetIdx: number
): number {
  const target = tape.get(targetIdx);
  const logPred = log(tape, predIdx);
  const oneMinusPredIdx = sub(tape, tape.track(constant(1)), predIdx);
  const logOneMinusPred = log(tape, oneMinusPredIdx);
  const term1 = mul(tape, targetIdx, logPred);
  const oneMinusTargetIdx = tape.track(constant(1 - target.data));
  const term2 = mul(tape, oneMinusTargetIdx, logOneMinusPred);
  return neg(tape, add(tape, term1, term2));
}

// ============================================================================
// Optimizer -- parameter update rules
// ============================================================================

export interface OptimizerConfig {
  learningRate: number;
  weightDecay?: number;
  momentum?: number;
}

export interface OptimizerState {
  config: OptimizerConfig;
  step: number;
  /** Per-parameter momentum buffer */
  velocities: Map<number, number>;
}

/**
 * Creates the Optimizer.
 */
export function createOptimizer(config: OptimizerConfig): OptimizerState {
  return {
    config,
    step: 0,
    velocities: new Map(),
  };
}

/**
 * SGD update with optional momentum and weight decay.
 */
export function optimizerStep(
  tape: GradientTape,
  optimizer: OptimizerState,
  paramIndices: number[]
): void {
  const { learningRate, weightDecay = 0, momentum = 0 } = optimizer.config;

  for (const idx of paramIndices) {
    const param = tape.get(idx);
    if (param.grad === null) continue;

    let grad = param.grad;

    // Weight decay (L2 regularization)
    if (weightDecay > 0) {
      grad += weightDecay * param.data;
    }

    // Momentum
    if (momentum > 0) {
      const prevV = optimizer.velocities.get(idx) ?? 0;
      const v = momentum * prevV + grad;
      optimizer.velocities.set(idx, v);
      grad = v;
    }

    param.data -= learningRate * grad;
  }

  optimizer.step++;
}

// ============================================================================
// Gradient analysis -- inspectable flow
// ============================================================================

export interface GradientFlowReport {
  /** Total parameters */
  paramCount: number;
  /** Parameters with non-zero gradient */
  activeGradients: number;
  /** Parameters with zero gradient (potentially blocked) */
  blockedGradients: number;
  /** Max absolute gradient */
  maxGrad: number;
  /** Mean absolute gradient */
  meanGrad: number;
  /** Whether vanishing gradient is detected */
  vanishing: boolean;
  /** Whether exploding gradient is detected */
  exploding: boolean;
}

/**
 * Handles the gnosis analyze Gradient Flow workflow.
 */
export function analyzeGradientFlow(tape: GradientTape): GradientFlowReport {
  const params = tape.parameters();
  const grads = params.map((p) => Math.abs(p.grad ?? 0));
  const nonZero = grads.filter((g) => g > 1e-12);

  const maxGrad = grads.length > 0 ? Math.max(...grads) : 0;
  const meanGrad =
    grads.length > 0 ? grads.reduce((a, b) => a + b, 0) / grads.length : 0;

  return {
    paramCount: params.length,
    activeGradients: nonZero.length,
    blockedGradients: params.length - nonZero.length,
    maxGrad,
    meanGrad,
    vanishing: meanGrad < 1e-7 && params.length > 0,
    exploding: maxGrad > 1e6,
  };
}
