// @a0n/gnosis/src/dockerfile-merkle.ts
/**
 * dockerfile-merkle.ts -- Build layer Merkle verification
 *
 * Bridges Docker build layer hashes into the void Merkle tree infrastructure.
 * Each layer in a Docker build produces a deterministic content-addressable
 * hash. This module converts those hashes into VoidVentRecords and builds
 * a VoidMerkleTree for tamper-evident build verification.
 *
 * Backed by PsycheGrindExtended.lean, Pass 18:
 *   - void_merkle_deterministic: same layers = same root
 *   - void_merkle_tamper_evident: modified layer changes root
 *   - fold_unfold_round_trip: exact round-trip
 *
 * @module @a0n/gnosis/dockerfile-merkle
 */

import type {
  VoidMerkleTree,
  VoidVentRecord,
  VoidMerkleVerification,
  VoidMerkleDiagnostics,
} from './void-merkle';
import {
  buildVoidMerkleTree,
  verifyVoidMerklePair,
  computeVoidMerkleDiagnostics,
} from './void-merkle';

// ============================================================================
// Types
// ============================================================================

/** A Docker build layer with its content-addressable hash. */
export interface BuildLayer {
  /** Layer index (0-based) */
  readonly index: number;
  /** Content-addressable hash (sha256) */
  readonly hash: string;
  /** Layer size in bytes */
  readonly sizeBytes: number;
  /** The Dockerfile instruction that produced this layer */
  readonly instruction?: string;
  /** Whether this layer was cached */
  readonly cached: boolean;
}

/** A complete Docker build record. */
export interface BuildRecord {
  /** The image tag */
  readonly imageTag: string;
  /** Build layers in order */
  readonly layers: readonly BuildLayer[];
  /** Build timestamp */
  readonly builtAt: number;
  /** The Merkle tree over build layers */
  readonly merkleTree: VoidMerkleTree;
  /** Diagnostics summary */
  readonly diagnostics: VoidMerkleDiagnostics;
}

/** Build reproducibility verification result. */
export interface BuildReproducibilityResult {
  /** Whether the builds are identical (same Merkle root) */
  readonly reproducible: boolean;
  /** The underlying Merkle pair verification */
  readonly merkleVerification: VoidMerkleVerification;
  /** First divergent layer index (-1 if identical) */
  readonly firstDivergentLayer: number;
  /** Number of matching layers */
  readonly matchingLayers: number;
  /** Total layers in build A */
  readonly totalLayersA: number;
  /** Total layers in build B */
  readonly totalLayersB: number;
}

// ============================================================================
// Construction
// ============================================================================

/**
 * Convert a layer hash string to a numeric magnitude for VoidVentRecord.
 * Uses the first 8 hex characters as a 32-bit integer.
 */
function hashToMagnitude(hash: string): number {
  const cleaned = hash.replace(/^sha256:/, '').slice(0, 8);
  return parseInt(cleaned, 16) || 0;
}

/**
 * Build a void Merkle tree from Docker build layers.
 *
 * Each layer becomes a VoidVentRecord where:
 *   - dimensionIdx = layer index
 *   - magnitude = hash-derived integer (first 32 bits of sha256)
 *   - sequenceNumber = layer index
 *
 * Mechanized: void_merkle_deterministic -- same layers = same root.
 */
export function buildLayerMerkleTree(
  layers: readonly BuildLayer[]
): VoidMerkleTree {
  const vents: VoidVentRecord[] = layers.map((layer) => ({
    dimensionIdx: layer.index,
    magnitude: hashToMagnitude(layer.hash),
    sequenceNumber: layer.index,
    timestamp: Date.now(),
  }));

  return buildVoidMerkleTree(vents);
}

/**
 * Create a complete build record from layers and image tag.
 */
export function createBuildRecord(
  imageTag: string,
  layers: readonly BuildLayer[]
): BuildRecord {
  const merkleTree = buildLayerMerkleTree(layers);
  const diagnostics = computeVoidMerkleDiagnostics(merkleTree);

  return {
    imageTag,
    layers,
    builtAt: Date.now(),
    merkleTree,
    diagnostics,
  };
}

/**
 * Create a BuildLayer from a hash string and optional metadata.
 */
export function createBuildLayer(
  index: number,
  hash: string,
  sizeBytes: number = 0,
  instruction?: string,
  cached: boolean = false
): BuildLayer {
  return { index, hash, sizeBytes, instruction, cached };
}

// ============================================================================
// Verification
// ============================================================================

/**
 * Verify build reproducibility by comparing two build records.
 *
 * Two builds of the same Dockerfile with the same context should produce
 * the same Merkle root. If they don't, the first divergent layer
 * identifies where non-determinism entered.
 *
 * Mechanized: void_merkle_tamper_evident, complement_preserved_until_new.
 */
export function verifyBuildReproducibility(
  buildA: BuildRecord,
  buildB: BuildRecord
): BuildReproducibilityResult {
  const merkleVerification = verifyVoidMerklePair(
    buildA.merkleTree,
    buildB.merkleTree
  );

  let firstDivergentLayer = -1;
  let matchingLayers = 0;
  const minLayers = Math.min(buildA.layers.length, buildB.layers.length);

  for (let i = 0; i < minLayers; i++) {
    if (buildA.layers[i].hash === buildB.layers[i].hash) {
      matchingLayers++;
    } else {
      firstDivergentLayer = i;
      break;
    }
  }

  if (
    firstDivergentLayer === -1 &&
    buildA.layers.length !== buildB.layers.length
  ) {
    firstDivergentLayer = minLayers;
  }

  return {
    reproducible: merkleVerification.rootsMatch,
    merkleVerification,
    firstDivergentLayer,
    matchingLayers,
    totalLayersA: buildA.layers.length,
    totalLayersB: buildB.layers.length,
  };
}

/**
 * Parse Docker build output to extract layer information.
 *
 * Handles both classic Docker build output and BuildKit output formats.
 */
export function parseBuildOutput(output: string): BuildLayer[] {
  const layers: BuildLayer[] = [];
  const lines = output.split('\n');

  // Classic format: "Step N/M : INSTRUCTION"
  // followed by "---> hash" or "CACHED"
  let currentInstruction: string | undefined;
  let layerIndex = 0;
  let nextIsCached = false;

  for (const line of lines) {
    const trimmed = line.trim();

    // Classic format step
    const stepMatch = trimmed.match(/^Step\s+\d+\/\d+\s*:\s*(.+)$/i);
    if (stepMatch) {
      currentInstruction = stepMatch[1];
      continue;
    }

    // Classic format cached indicator (appears before the hash line)
    if (trimmed === 'Using cache' || trimmed === 'CACHED') {
      nextIsCached = true;
      continue;
    }

    // Classic format layer hash (with or without sha256: prefix)
    const hashMatch = trimmed.match(/^--->\s+(?:sha256:)?([0-9a-f]{12,64})$/);
    if (hashMatch) {
      layers.push(
        createBuildLayer(
          layerIndex++,
          hashMatch[1],
          0,
          currentInstruction,
          nextIsCached
        )
      );
      currentInstruction = undefined;
      nextIsCached = false;
      continue;
    }

    // BuildKit format: "#N [stage N/M] INSTRUCTION"
    const buildkitMatch = trimmed.match(/^#\d+\s+\[.*?\]\s+(.+)$/);
    if (buildkitMatch) {
      currentInstruction = buildkitMatch[1];
      nextIsCached = trimmed.includes('CACHED');
      continue;
    }

    // BuildKit digest line: "digest: sha256:..."
    const digestMatch = trimmed.match(/digest:\s+(sha256:[0-9a-f]{64})/);
    if (digestMatch) {
      layers.push(
        createBuildLayer(
          layerIndex++,
          digestMatch[1],
          0,
          currentInstruction,
          nextIsCached
        )
      );
      currentInstruction = undefined;
      nextIsCached = false;
    }
  }

  return layers;
}
