// @a0n/gnosis/src/doppler-redshift.ts
/**
 * Doppler Redshift — relativistic Doppler shift and cosmological redshift.
 *
 * Computes the observed wavelength shift of light from moving or
 * distant sources. Two independent effects contribute:
 *
 *   Kinematic (Doppler): relative motion between source and observer.
 *   Cosmological (Hubble): expansion of space stretching wavelengths.
 *
 * Both effects produce a dimensionless redshift parameter z:
 *   z > 0  →  redshift (receding / distant)
 *   z < 0  →  blueshift (approaching)
 *   z = 0  →  no shift
 *
 * The combined redshift multiplies the individual factors:
 *   (1 + z_total) = (1 + z_kin) · (1 + z_cosmo)
 *
 * @packageDocumentation
 */

import {
  SPEED_OF_LIGHT_MS,
  HUBBLE_CONSTANT_KM_S_MPC,
  MPC_IN_LY,
} from './math';

import type { SpacetimeEvent } from './causal-topology';

// ─── Types ─────────────────────────────────────────────────────────

export type RedshiftSource = 'kinematic' | 'cosmological' | 'combined';
export type RedshiftColorScheme = 'physical' | 'exaggerated';

export interface RedshiftColor {
  readonly r: number;
  readonly g: number;
  readonly b: number;
}

export interface RedshiftResult {
  readonly bodyId: string;
  readonly label: string;
  readonly redshiftZ: number;
  readonly source: RedshiftSource;
  readonly radialVelocityKms?: number;
  readonly distanceLy: number;
  readonly colorTint: RedshiftColor;
}

export interface DopplerShiftedEvent {
  readonly event: SpacetimeEvent;
  readonly redshiftZ: number;
  readonly source: RedshiftSource;
  readonly colorTint: RedshiftColor;
}

// ─── Constants ─────────────────────────────────────────────────────

/** Speed of light in km/s. */
const SPEED_OF_LIGHT_KMS = SPEED_OF_LIGHT_MS / 1000;

// ─── Pure Functions ────────────────────────────────────────────────

/**
 * Convert a velocity in km/s to a fraction of the speed of light (β).
 */
export function kmsToFractionC(kms: number): number {
  return kms / SPEED_OF_LIGHT_KMS;
}

/**
 * Relativistic (special-relativistic) Doppler redshift.
 *
 * z = √((1 + β) / (1 - β)) - 1
 *
 * Positive β means the source recedes (redshift, z > 0).
 * Negative β means the source approaches (blueshift, z < 0).
 *
 * β must satisfy |β| < 1 (sub-luminal).
 */
export function kinematicRedshift(beta: number): number {
  if (Math.abs(beta) >= 1) {
    throw new RangeError(`|β| must be < 1, got ${beta}`);
  }
  return Math.sqrt((1 + beta) / (1 - beta)) - 1;
}

/**
 * First-order cosmological redshift from the Hubble flow.
 *
 * v = H₀ · d_Mpc   (km/s)
 * z = v / c
 *
 * Valid for z ≪ 1 (nearby universe).
 */
export function cosmologicalRedshift(
  distanceLy: number,
  hubbleKmsMpc: number = HUBBLE_CONSTANT_KM_S_MPC,
): number {
  const distanceMpc = distanceLy / MPC_IN_LY;
  const recessionalVelocityKms = hubbleKmsMpc * distanceMpc;
  return recessionalVelocityKms / SPEED_OF_LIGHT_KMS;
}

/**
 * Combine independent redshift contributions.
 *
 * (1 + z_total) = (1 + z_kin) · (1 + z_cosmo)
 */
export function combinedRedshift(zKin: number, zCosmo: number): number {
  return (1 + zKin) * (1 + zCosmo) - 1;
}

/**
 * Map a redshift value to an RGB color tint.
 *
 * Base neutral is white (255, 255, 255).
 *   z > 0 → shift toward red (increase r, decrease b).
 *   z < 0 → shift toward blue (increase b, decrease r).
 *
 * 'physical' uses subtle tints (±50 max shift).
 * 'exaggerated' uses dramatic colors (±200 max shift).
 */
export function redshiftColor(
  z: number,
  scheme: RedshiftColorScheme = 'physical',
): RedshiftColor {
  const multiplier = scheme === 'exaggerated' ? 200 : 50;
  const maxShift = scheme === 'exaggerated' ? 200 : 50;
  const shift = Math.max(-maxShift, Math.min(maxShift, z * multiplier));

  return {
    r: Math.round(Math.max(0, Math.min(255, 255 + shift))),
    g: 255,
    b: Math.round(Math.max(0, Math.min(255, 255 - shift))),
  };
}

/**
 * Compute the full redshift result for a celestial body.
 *
 * Combines kinematic redshift (if radial velocity is provided) with
 * cosmological redshift from the Hubble flow.
 */
export function computeRedshift(
  bodyId: string,
  label: string,
  distanceLy: number,
  radialVelocityKms?: number,
): RedshiftResult {
  const zCosmo = cosmologicalRedshift(distanceLy);

  let zKin = 0;
  let hasKinematic = false;
  if (radialVelocityKms !== undefined) {
    const beta = kmsToFractionC(radialVelocityKms);
    zKin = kinematicRedshift(beta);
    hasKinematic = true;
  }

  const source: RedshiftSource = hasKinematic
    ? zCosmo !== 0
      ? 'combined'
      : 'kinematic'
    : 'cosmological';

  const redshiftZ = hasKinematic ? combinedRedshift(zKin, zCosmo) : zCosmo;

  return {
    bodyId,
    label,
    redshiftZ,
    source,
    ...(radialVelocityKms !== undefined ? { radialVelocityKms } : {}),
    distanceLy,
    colorTint: redshiftColor(redshiftZ),
  };
}
