// @a0n/gnosis/src/echolocation.ts
/**
 * echolocation.ts -- Reconstruct a person's void boundary from micro-interactions
 *
 * The idea: if you ping someone enough times at the fast timescale (seconds),
 * the rejection pattern at the shallow layer projects onto the deep structure.
 * Chester's health anxiety (years) shows up as rejection of low offers (seconds).
 * After 500 rounds, the $100K-$120K zone has 214-259 rejections. That density
 * formalizes the health anxiety, measured from the outside.
 *
 * Echolocation runs many small, low-stakes probes across all dimensions of the
 * action space. Each probe is a micro-interaction: present a stimulus, observe
 * acceptance or rejection, accumulate. The accumulated void boundary formalizes the
 * personality -- at whatever resolution the probe density achieves.
 *
 * The BoundaryStack's upward constraint flow means shallow observations
 * reconstruct deep structure. The downward context flow means once you've
 * identified the deep structure, it predicts shallow behavior. The echo
 * becomes a map.
 */

import {
  type VoidBoundary,
  type BoundaryStack,
  type TimescaleBoundary,
  type Measurement,
  type Gait,
  createVoidBoundary,
  createTimescaleBoundary,
  createBoundaryStack,
  createResonance,
  complementDistribution,
  updateVoidBoundary,
  shannonEntropy,
  excessKurtosis,
  measure,
  flattenStack,
  measureStack,
  tickBoundaryStack,
  upwardConstraint,
} from './void';

import {
  type VoidAgent,
  type PersonalityLayerConfig,
  createVoidAgent,
  step,
  personalityVector,
  metacogState,
} from './void-agent';

// ============================================================================
// Probe Types
// ============================================================================

/**
 * A Probe is a micro-interaction designed to elicit a response.
 * The response (accept/reject + magnitude) fills the echo boundary.
 */
export interface Probe {
  /** Which dimension of the action space this probe targets */
  dimension: number;
  /** Optional label for introspection */
  label?: string;
  /** Intensity: how much rejection signal a negative response produces */
  intensity: number;
}

/**
 * A ProbeResponse is what the subject returns when probed.
 * This is the raw data the echolocation engine accumulates.
 */
export interface ProbeResponse {
  probe: Probe;
  /** Whether the subject rejected the probe */
  rejected: boolean;
  /** Magnitude of rejection (0 = neutral, higher = stronger rejection) */
  magnitude: number;
  /** Optional qualitative signal (valence, affect, etc.) */
  valence?: number;
}

// ============================================================================
// Probe Generation Strategies
// ============================================================================

/**
 * Uniform sweep: probe every dimension equally.
 * Best for initial exploration when nothing is known.
 */
export function uniformProbes(
  dimensions: number,
  intensity: number = 1.0,
  labels?: readonly string[]
): Probe[] {
  return Array.from({ length: dimensions }, (_, i) => ({
    dimension: i,
    label: labels?.[i],
    intensity,
  }));
}

/**
 * Complement-guided probes: probe dimensions where the echo
 * boundary has the LEAST void (most uncertain regions).
 * Best for targeted exploration after initial sweep.
 */
export function complementGuidedProbes(
  echoBoundary: VoidBoundary,
  eta: number,
  count: number,
  intensity: number = 1.0
): Probe[] {
  const dist = complementDistribution(echoBoundary, eta);
  // Sort by complement weight (highest = least explored)
  const ranked = dist
    .map((w, i) => ({ dim: i, weight: w }))
    .sort((a, b) => b.weight - a.weight);

  return ranked.slice(0, count).map((r) => ({
    dimension: r.dim,
    intensity,
  }));
}

/**
 * Neighborhood probes: given a known high-void dimension,
 * probe its neighbors to map the rejection boundary shape.
 */
export function neighborhoodProbes(
  center: number,
  radius: number,
  totalDimensions: number,
  intensity: number = 1.0
): Probe[] {
  const probes: Probe[] = [];
  for (let d = -radius; d <= radius; d++) {
    const dim = center + d;
    if (dim >= 0 && dim < totalDimensions) {
      probes.push({ dimension: dim, intensity: intensity / (1 + Math.abs(d)) });
    }
  }
  return probes;
}

// ============================================================================
// Echo Boundary -- the accumulated echo of the subject's void
// ============================================================================

export interface EchoBoundary {
  /** The reconstructed void boundary */
  boundary: VoidBoundary;
  /** Per-dimension probe count (how many times we've pinged each dimension) */
  probeCounts: number[];
  /** Per-dimension rejection rate (rejections / probes) */
  rejectionRates: number[];
  /** Total probes sent */
  totalProbes: number;
  /** Total rejections observed */
  totalRejections: number;
  /** Dimension labels (if provided) */
  labels?: readonly string[];
}

/**
 * Creates the Echo Boundary.
 */
export function createEchoBoundary(
  dimensions: number,
  labels?: readonly string[]
): EchoBoundary {
  return {
    boundary: createVoidBoundary(dimensions),
    probeCounts: new Array(dimensions).fill(0),
    rejectionRates: new Array(dimensions).fill(0),
    totalProbes: 0,
    totalRejections: 0,
    labels,
  };
}

/** Accumulate a probe response into the echo boundary */
export function accumulateEcho(
  echo: EchoBoundary,
  response: ProbeResponse
): void {
  const dim = response.probe.dimension;
  echo.probeCounts[dim]++;
  echo.totalProbes++;

  if (response.rejected) {
    updateVoidBoundary(echo.boundary, dim, response.magnitude);
    echo.totalRejections++;
  }

  // Update rejection rate
  echo.rejectionRates[dim] =
    echo.probeCounts[dim] > 0
      ? echo.boundary.counts[dim] / echo.probeCounts[dim]
      : 0;
}

/** Measure the echo boundary -- how much have we learned? */
export function measureEcho(
  echo: EchoBoundary,
  eta: number = 3.0
): {
  measurement: Measurement;
  coverage: number;
  confidence: number;
} {
  const m = measure(echo.boundary, eta, echo.totalProbes);
  const dimensions = echo.probeCounts.length;
  const probedDims = echo.probeCounts.filter((c) => c > 0).length;
  const coverage = probedDims / dimensions;
  // Confidence: how concentrated is the void? Higher kurtosis = more certain
  const confidence = Math.min(1, Math.max(0, (m.kurtosis + 3) / 6));
  return { measurement: m, coverage, confidence };
}

// ============================================================================
// Echolocation Engine
// ============================================================================

export interface EcholocationConfig {
  /** Number of dimensions in the action space */
  dimensions: number;
  /** Dimension labels */
  labels?: readonly string[];
  /** Initial probe intensity */
  intensity?: number;
  /** Temperature for complement-guided probing */
  eta?: number;
  /** Probes per sweep round */
  probesPerRound?: number;
  /** Maximum total probes */
  maxProbes?: number;
  /** Convergence threshold: stop when entropy change < this */
  convergenceThreshold?: number;
}

export interface EcholocationResult {
  /** The reconstructed echo boundary */
  echo: EchoBoundary;
  /** The personality profile extracted from the echo */
  profile: number[];
  /** Per-dimension rejection rates (the raw personality signal) */
  rejectionRates: number[];
  /** Rounds completed */
  rounds: number;
  /** Total probes sent */
  totalProbes: number;
  /** Whether the echolocation converged */
  converged: boolean;
  /** Entropy history (should decrease as we learn more) */
  entropyHistory: number[];
  /** Measurement at convergence */
  measurement: Measurement;
}

/**
 * Run echolocation against a subject.
 *
 * The subject is a function that takes a probe and returns a response.
 * It can be a live person (via UI), a VoidAgent (for testing), or
 * any oracle that maps stimuli to acceptance/rejection.
 *
 * The engine:
 * 1. Starts with a uniform sweep (probe every dimension)
 * 2. Switches to complement-guided probing (target uncertain areas)
 * 3. Converges when entropy stabilizes (we've learned what we can)
 *
 * The resulting echo boundary formalizes the person's void, measured from outside.
 */
export function echolocate(
  config: EcholocationConfig,
  subject: (probe: Probe) => ProbeResponse
): EcholocationResult {
  const eta = config.eta ?? 3.0;
  const intensity = config.intensity ?? 1.0;
  const probesPerRound = config.probesPerRound ?? config.dimensions;
  const maxProbes = config.maxProbes ?? config.dimensions * 50;
  const convergenceThreshold = config.convergenceThreshold ?? 0.01;

  const echo = createEchoBoundary(config.dimensions, config.labels);
  const entropyHistory: number[] = [];
  let round = 0;
  let converged = false;

  // Phase 1: uniform sweep (at least one probe per dimension)
  const initialProbes = uniformProbes(
    config.dimensions,
    intensity,
    config.labels
  );
  for (const probe of initialProbes) {
    const response = subject(probe);
    accumulateEcho(echo, response);
  }

  const initialMeasure = measureEcho(echo, eta);
  entropyHistory.push(initialMeasure.measurement.entropy);
  round++;

  // Phase 2: complement-guided probing until convergence
  while (echo.totalProbes < maxProbes && !converged) {
    const probes = complementGuidedProbes(
      echo.boundary,
      eta,
      probesPerRound,
      intensity
    );
    for (const probe of probes) {
      const response = subject(probe);
      accumulateEcho(echo, response);
    }

    const m = measureEcho(echo, eta);
    entropyHistory.push(m.measurement.entropy);
    round++;

    // Convergence check: entropy change is small
    if (entropyHistory.length >= 3) {
      const recent = entropyHistory.slice(-3);
      const maxDelta = Math.max(
        Math.abs(recent[2] - recent[1]),
        Math.abs(recent[1] - recent[0])
      );
      if (maxDelta < convergenceThreshold) {
        converged = true;
      }
    }
  }

  const finalMeasure = measureEcho(echo, eta);
  const profile = complementDistribution(echo.boundary, eta);

  return {
    echo,
    profile,
    rejectionRates: [...echo.rejectionRates],
    rounds: round,
    totalProbes: echo.totalProbes,
    converged,
    entropyHistory,
    measurement: finalMeasure.measurement,
  };
}

// ============================================================================
// Echo → BoundaryStack projection (echolocation → personality)
// ============================================================================

/**
 * Project an echo boundary onto the 7-layer personality stack.
 *
 * The echo is measured at the `seconds` timescale (fast, shallow).
 * This function distributes the echo signal across personality layers
 * using the same magnitude attenuation that propagateToPersonality uses:
 *   - Behaviors (months): 80% of echo signal
 *   - Mental health (weeks): 150% for high-rejection dims, 30% otherwise
 *   - Traits (years): 10%
 *   - Temperament (lifetime): 1%
 *   - Culture (generational): 1%
 *
 * The resulting stack formalizes the personality, reconstructed from echoes.
 */
export function echoToPersonality(
  echo: EchoBoundary,
  layerConfigs?: PersonalityLayerConfig[]
): BoundaryStack {
  const dims = echo.boundary.counts.length;

  const defaultLayers: PersonalityLayerConfig[] = layerConfigs ?? [
    { name: 'temperament', timescale: 'lifetime', dimensions: dims },
    { name: 'attachment', timescale: 'lifetime', dimensions: dims },
    { name: 'traits', timescale: 'years', dimensions: dims },
    { name: 'behaviors', timescale: 'months', dimensions: dims },
    { name: 'mental-health', timescale: 'weeks', dimensions: dims },
    { name: 'history', timescale: 'years', dimensions: dims },
    { name: 'culture', timescale: 'generational', dimensions: dims },
  ];

  const layers: TimescaleBoundary[] = defaultLayers.map((lc) => {
    const tb = createTimescaleBoundary(
      lc.name,
      lc.timescale,
      lc.dimensions,
      lc.labels
    );

    // Project echo signal with timescale-appropriate attenuation
    const attenuation = getAttenuation(lc.timescale);
    for (let i = 0; i < Math.min(dims, lc.dimensions); i++) {
      const echoDensity = echo.boundary.counts[i];
      const isHighRejection = echo.rejectionRates[i] > 0.5;

      let magnitude = echoDensity * attenuation;
      // Mental health amplifies high-rejection dimensions
      if (lc.name === 'mental-health' && isHighRejection) {
        magnitude = echoDensity * 1.5;
      }

      if (magnitude > 0) {
        updateVoidBoundary(tb.boundary, i, magnitude);
      }
    }
    return tb;
  });

  const resonances = [
    createResonance(0, 4, 0.05, 'temperament <-> mental health'),
    createResonance(6, 1, 0.05, 'culture <-> attachment'),
    createResonance(2, 3, 0.03, 'traits <-> behaviors'),
  ];

  return createBoundaryStack('echolocated-personality', layers, resonances);
}

function getAttenuation(timescale: string): number {
  switch (timescale) {
    case 'instant':
    case 'seconds':
    case 'minutes':
      return 1.0;
    case 'hours':
    case 'days':
      return 0.5;
    case 'weeks':
      return 0.3;
    case 'months':
      return 0.8;
    case 'years':
      return 0.1;
    case 'lifetime':
      return 0.01;
    case 'generational':
      return 0.01;
    default:
      return 0.1;
  }
}

// ============================================================================
// Digital Twin -- the echolocated personality as a VoidAgent
// ============================================================================

/**
 * Create a digital twin from an echolocation result.
 *
 * The twin is a VoidAgent whose personality stack is the echolocated
 * boundary, projected onto the 7-layer model. It walks the same void.
 * Its complement distribution matches the subject's rejection patterns.
 * Its gait adapts to the same kurtosis. It formalizes the person, measured
 * from the outside and reconstructed on the inside.
 */
export function createDigitalTwin(
  name: string,
  echoResult: EcholocationResult,
  rng: () => number = Math.random
): VoidAgent {
  const personality = echoToPersonality(echoResult.echo);
  return createVoidAgent(
    {
      name: `${name}:twin`,
      actionDimensions: echoResult.echo.boundary.counts.length,
      personality,
      eta: 2.0,
      numHeads: 2,
      neighborhoodRadius: 1,
      decayRate: 0.005,
    },
    rng
  );
}

// ============================================================================
// Fidelity -- how well does the twin match the original?
// ============================================================================

/**
 * Measure twin fidelity: run the same probes against the original
 * and the twin, compare their response patterns.
 *
 * Returns L1 distance between the two complement distributions
 * (0 = perfect match, 2 = maximally different).
 */
export function measureFidelity(
  original: (probe: Probe) => ProbeResponse,
  twin: VoidAgent,
  probes: Probe[]
): {
  l1Distance: number;
  perDimensionDelta: number[];
  matchRate: number;
} {
  const twinDist = complementDistribution(
    twin.perception[0].walker.boundary,
    twin.perception[0].walker.eta
  );

  // Run probes against original, build echo
  const origEcho = createEchoBoundary(twin.actionDimensions);
  for (const probe of probes) {
    accumulateEcho(origEcho, original(probe));
  }
  const origDist = complementDistribution(
    origEcho.boundary,
    twin.perception[0].walker.eta
  );

  // L1 distance
  let l1 = 0;
  const deltas: number[] = [];
  let matches = 0;
  for (let i = 0; i < origDist.length; i++) {
    const delta = Math.abs(origDist[i] - (twinDist[i] ?? 0));
    l1 += delta;
    deltas.push(delta);
    // Match if both rank this dimension similarly (within 10% of max)
    if (delta < 0.1) matches++;
  }

  return {
    l1Distance: l1,
    perDimensionDelta: deltas,
    matchRate: matches / origDist.length,
  };
}
