// @a0n/gnosis/src/ecology-graph-snapshot.ts
/**
 * Versioned ecological graph export (bounded pilot). Pure structural validation.
 * See docs/ECOLOGY_PILOT_SPEC.md — this is a data contract, not a scientific claim.
 */

import type { VerifyResult } from './verify-bundle';

export type EcologyEdgeProvenance = 'sensor' | 'survey' | 'model';

export interface EcologyNodeV1 {
  readonly id: string;
  readonly taxon?: string;
}

export interface EcologyEdgeV1 {
  readonly u: string;
  readonly v: string;
  readonly weight?: number;
  readonly weight_ci?: readonly [number, number];
  readonly provenance: EcologyEdgeProvenance;
}

export interface EcologyGraphSnapshotV1 {
  readonly schemaVersion: '1.0.0';
  readonly domain: string;
  readonly graph_id: string;
  readonly exported_at: string;
  readonly nodes: readonly EcologyNodeV1[];
  readonly edges: readonly EcologyEdgeV1[];
}

/**
 * Handles the gnosis validate Ecology Graph Snapshot workflow.
 */
export function validateEcologyGraphSnapshot(x: unknown): VerifyResult {
  if (x === null || typeof x !== 'object') {
    return { ok: false, reason: 'not_object' };
  }
  const o = x as Record<string, unknown>;
  if (o.schemaVersion !== '1.0.0') {
    return { ok: false, reason: 'schema_version' };
  }
  if (typeof o.domain !== 'string' || o.domain.length < 1) {
    return { ok: false, reason: 'domain' };
  }
  if (typeof o.graph_id !== 'string' || o.graph_id.length < 4) {
    return { ok: false, reason: 'graph_id' };
  }
  if (typeof o.exported_at !== 'string' || Number.isNaN(Date.parse(o.exported_at))) {
    return { ok: false, reason: 'exported_at' };
  }
  const nodes = o.nodes;
  if (!Array.isArray(nodes) || nodes.length < 1) {
    return { ok: false, reason: 'nodes' };
  }
  const ids = new Set<string>();
  for (let i = 0; i < nodes.length; i++) {
    const n = nodes[i];
    if (n === null || typeof n !== 'object') {
      return { ok: false, reason: `node_${i}` };
    }
    const nr = n as Record<string, unknown>;
    if (typeof nr.id !== 'string' || nr.id.length < 1) {
      return { ok: false, reason: `node_${i}_id` };
    }
    if (ids.has(nr.id)) {
      return { ok: false, reason: 'duplicate_node_id' };
    }
    ids.add(nr.id);
    if (nr.taxon !== undefined && typeof nr.taxon !== 'string') {
      return { ok: false, reason: `node_${i}_taxon` };
    }
  }
  const edges = o.edges;
  if (!Array.isArray(edges)) {
    return { ok: false, reason: 'edges' };
  }
  for (let j = 0; j < edges.length; j++) {
    const e = edges[j];
    if (e === null || typeof e !== 'object') {
      return { ok: false, reason: `edge_${j}` };
    }
    const er = e as Record<string, unknown>;
    if (typeof er.u !== 'string' || !ids.has(er.u)) {
      return { ok: false, reason: `edge_${j}_u` };
    }
    if (typeof er.v !== 'string' || !ids.has(er.v)) {
      return { ok: false, reason: `edge_${j}_v` };
    }
    if (
      er.weight !== undefined &&
      (typeof er.weight !== 'number' || !Number.isFinite(er.weight))
    ) {
      return { ok: false, reason: `edge_${j}_weight` };
    }
    const wci = er.weight_ci;
    if (wci !== undefined) {
      if (!Array.isArray(wci) || wci.length !== 2) {
        return { ok: false, reason: `edge_${j}_weight_ci` };
      }
      if (typeof wci[0] !== 'number' || typeof wci[1] !== 'number') {
        return { ok: false, reason: `edge_${j}_weight_ci_numeric` };
      }
    }
    if (er.provenance !== 'sensor' && er.provenance !== 'survey' && er.provenance !== 'model') {
      return { ok: false, reason: `edge_${j}_provenance` };
    }
  }
  return { ok: true };
}
