// @a0n/gnosis/src/eyetracking.test.ts
import { describe, expect, test } from '@a0n/gnosis/test';

import {
  appendEyeTrackingBuleyeanRejectionBatch,
  appendEyeTrackingObservation,
  auditEyeTrackingSession,
  createEyeTrackingBuleyeanTrainingJob,
  createEyeTrackingDatasetSlice,
  createEyeTrackingQDoc,
  createEyeTrackingSession,
  createEyeTrackingTrainingManifest,
  exportEyeTrackingBuleyeanPersonalizationDelta,
  loadEyeTrackingSessionByAliasFromQDoc,
  recordEyeTrackingCalibrationSample,
  recordEyeTrackingFailure,
  selectEyeTrackingSessions,
  serializeEyeTrackingRejectionJSONL,
  storeEyeTrackingSessionInQDoc,
  summarizeEyeTrackingBuleyeanTrainingCorpus,
  summarizeEyeTrackingFailurePressure,
  type EyeTrackingCalibrationPlan,
} from './eyetracking';

const TEST_PLAN: EyeTrackingCalibrationPlan = {
  points: [{ id: 'center', x: 0.5, y: 0.5 }],
  samplesPerPoint: 1,
  maxAttemptsPerPoint: 2,
};

describe('eyetracking corpus', () => {
  test('stores sessions in QDoc and audits the failure surface', () => {
    const session = createEyeTrackingSession({
      sessionId: 'session-1',
      calibrationPlan: TEST_PLAN,
      deviceId: 'camera-a',
      modelVersion: 'aeon-v1',
    });

    const observation = appendEyeTrackingObservation(session, {
      timestamp: 1,
      left: {
        pupil: { x: 0.49, y: 0.48 },
        openness: 0.9,
        quality: 0.92,
      },
      right: {
        pupil: { x: 0.51, y: 0.48 },
        openness: 0.91,
        quality: 0.94,
      },
      gazePoint: { x: 0.5, y: 0.5 },
      confidence: 0.93,
      blink: false,
      eyeBlocked: false,
      fixationDurationMs: 160,
      saccadeVelocity: 0.2,
      scanningVariance: 0.08,
      faceQuality: 0.95,
      source: 'replay',
    });

    recordEyeTrackingCalibrationSample(session, {
      pointId: 'center',
      observationId: observation.observationId,
      timestamp: 2,
      quality: 0.95,
    });
    recordEyeTrackingFailure(session, {
      timestamp: 3,
      kind: 'zone-miss',
      severity: 'error',
      message: 'Predicted top-left for a centered target.',
      observationId: observation.observationId,
      expectedZone: 'center',
      observedZone: 'top-left',
    });

    const audit = auditEyeTrackingSession(session);
    expect(session.status).toBe('tracking');
    expect(audit.pass).toBe(true);
    expect(audit.merkle.size).toBe(1);

    const doc = createEyeTrackingQDoc('eyetracking-test');
    storeEyeTrackingSessionInQDoc(doc, session, ['latest']);
    const restored = loadEyeTrackingSessionByAliasFromQDoc(doc, 'latest');

    expect(restored?.sessionId).toBe('session-1');
    expect(restored?.failures.length).toBe(1);
    expect(restored?.calibration.ready).toBe(true);
  });

  test('summarizes failure pressure and serializes rejection JSONL', () => {
    const session = createEyeTrackingSession({
      sessionId: 'session-2',
      calibrationPlan: TEST_PLAN,
      deviceId: 'camera-b',
      modelVersion: 'aeon-v2',
    });

    recordEyeTrackingFailure(session, {
      timestamp: 10,
      kind: 'low-quality',
      severity: 'warning',
      message: 'Lighting collapsed below the acceptance floor.',
    });
    recordEyeTrackingFailure(session, {
      timestamp: 11,
      kind: 'low-quality',
      severity: 'warning',
      message: 'Lighting collapsed below the acceptance floor.',
    });

    const pressure = summarizeEyeTrackingFailurePressure(session);
    const jsonl = serializeEyeTrackingRejectionJSONL([session]);

    expect(pressure.byKind['low-quality']).toBe(4);
    expect(pressure.byDevice['camera-b']).toBe(4);
    expect(pressure.hottestKind).toBe('low-quality');
    expect(jsonl).toContain('eyetracking:session-2');
    expect(jsonl).toContain('low-quality');
  });

  test('builds dataset slices and training manifests for ETaaS control planes', () => {
    const trainSession = createEyeTrackingSession({
      sessionId: 'session-train',
      calibrationPlan: TEST_PLAN,
      deviceId: 'camera-a',
      modelVersion: 'aeon-v3',
      startedAt: 100,
      status: 'tracking',
    });
    recordEyeTrackingFailure(trainSession, {
      timestamp: 101,
      kind: 'low-quality',
      severity: 'warning',
      message: 'Low-light frame drifted below the acceptance floor.',
    });

    const evalSession = createEyeTrackingSession({
      sessionId: 'session-eval',
      calibrationPlan: TEST_PLAN,
      deviceId: 'camera-b',
      modelVersion: 'aeon-v3',
      startedAt: 200,
      status: 'tracking',
    });
    recordEyeTrackingFailure(evalSession, {
      timestamp: 201,
      kind: 'zone-miss',
      severity: 'error',
      message: 'Center target resolved to top-center.',
    });

    const selected = selectEyeTrackingSessions(
      [trainSession, evalSession],
      {
        deviceIds: ['camera-b'],
        failureKinds: ['zone-miss'],
      }
    );
    const evalSlice = createEyeTrackingDatasetSlice(
      [trainSession, evalSession],
      {
        split: 'eval',
        selection: {
          deviceIds: ['camera-b'],
        },
        createdAt: 300,
      }
    );
    const evalSliceRepeat = createEyeTrackingDatasetSlice(
      [trainSession, evalSession],
      {
        split: 'eval',
        selection: {
          deviceIds: ['camera-b'],
        },
        createdAt: 300,
      }
    );
    const manifest = createEyeTrackingTrainingManifest(
      [
        {
          split: 'train',
          source: [trainSession],
        },
        {
          split: 'eval',
          source: [trainSession, evalSession],
          selection: { deviceIds: ['camera-b'] },
        },
      ],
      {
        generatedAt: 400,
        metadata: { deployment: 'dash-lab' },
      }
    );
    const manifestRepeat = createEyeTrackingTrainingManifest(
      [
        {
          split: 'train',
          source: [trainSession],
        },
        {
          split: 'eval',
          source: [trainSession, evalSession],
          selection: { deviceIds: ['camera-b'] },
        },
      ],
      {
        generatedAt: 400,
        metadata: { deployment: 'dash-lab' },
      }
    );

    expect(selected.map((session) => session.sessionId)).toEqual([
      'session-eval',
    ]);
    expect(evalSlice.sliceId).toBe(evalSliceRepeat.sliceId);
    expect(evalSlice.sliceId).toMatch(/^eye-slice:eval:/);
    expect(evalSlice.sessionIds).toEqual(['session-eval']);
    expect(evalSlice.pressure.hottestKind).toBe('zone-miss');
    expect(evalSlice.rejectionJsonl).toContain('eyetracking:session-eval');
    expect(manifest.service).toBe('eyetracking-as-a-service');
    expect(manifest.manifestId).toBe(manifestRepeat.manifestId);
    expect(manifest.manifestId).toMatch(/^eye-manifest:/);
    expect(manifest.totalSessionCount).toBe(2);
    expect(manifest.hottestKinds).toEqual(['zone-miss', 'low-quality']);
    expect(manifest.metadata).toEqual({ deployment: 'dash-lab' });
  });

  test('builds Buleyean jobs, appends rejection batches, and exports personalization deltas', () => {
    const trainSession = createEyeTrackingSession({
      sessionId: 'session-bule-train',
      calibrationPlan: TEST_PLAN,
      deviceId: 'camera-a',
      modelVersion: 'aeon-v4',
      startedAt: 1000,
      status: 'tracking',
    });
    recordEyeTrackingFailure(trainSession, {
      timestamp: 1001,
      kind: 'low-quality',
      severity: 'warning',
      message: 'Low-light frame drifted below the acceptance floor.',
    });

    const evalSession = createEyeTrackingSession({
      sessionId: 'session-bule-eval',
      calibrationPlan: TEST_PLAN,
      deviceId: 'camera-b',
      modelVersion: 'aeon-v4',
      startedAt: 1100,
      status: 'completed',
    });
    recordEyeTrackingFailure(evalSession, {
      timestamp: 1101,
      kind: 'zone-miss',
      severity: 'error',
      message: 'Center target resolved to top-center.',
    });

    const personalizationSession = createEyeTrackingSession({
      sessionId: 'session-bule-personal',
      calibrationPlan: TEST_PLAN,
      deviceId: 'camera-c',
      modelVersion: 'aeon-v5',
      startedAt: 1200,
      status: 'failed',
    });
    recordEyeTrackingFailure(personalizationSession, {
      timestamp: 1201,
      kind: 'occlusion',
      severity: 'fatal',
      message: 'Face disappeared behind the tablet edge.',
    });

    const corpus = summarizeEyeTrackingBuleyeanTrainingCorpus(
      [trainSession, evalSession]
    );
    const job = createEyeTrackingBuleyeanTrainingJob(
      [
        {
          split: 'train',
          source: [trainSession],
        },
        {
          split: 'eval',
          source: [trainSession, evalSession],
          selection: { deviceIds: ['camera-b'] },
        },
      ],
      {
        createdAt: 1300,
        metadata: { deployment: 'dash-lab' },
      }
    );
    const appended = appendEyeTrackingBuleyeanRejectionBatch(
      job,
      {
        split: 'shadow',
        source: [trainSession, personalizationSession],
      },
      {
        createdAt: 1400,
      }
    );
    const delta = exportEyeTrackingBuleyeanPersonalizationDelta(
      [trainSession, evalSession, personalizationSession],
      {
        createdAt: 1500,
        selection: { deviceIds: ['camera-c'] },
        metadata: { subjectId: 'user-1' },
      }
    );

    expect(corpus.promptCount).toBe(2);
    expect(corpus.totalRejectedResponses).toBe(2);
    expect(corpus.totalRejectionMass).toBe(5);
    expect(corpus.byStatus.tracking).toBe(1);
    expect(corpus.byStatus.completed).toBe(1);

    expect(job.trainer).toBe('buleyean-rl');
    expect(job.jobId).toMatch(/^eye-buleyean-job:/);
    expect(job.corpus.totalRejectionMass).toBe(5);
    expect(job.rejectionJsonl).toContain('eyetracking:session-bule-train');
    expect(job.splits.map((split) => split.split)).toEqual(['train', 'eval']);
    expect(job.config.federated.dimensions).toBe(12);

    expect(appended.duplicateSessionIds).toEqual(['session-bule-train']);
    expect(appended.appendedSessionIds).toEqual(['session-bule-personal']);
    expect(appended.job.corpus.totalRejectionMass).toBe(10);
    expect(appended.job.metadata).toEqual({
      deployment: 'dash-lab',
      parentJobId: job.jobId,
    });

    expect(delta.deltaId).toMatch(/^eye-buleyean-delta:/);
    expect(delta.sessionIds).toEqual(['session-bule-personal']);
    expect(delta.corpus.hottestKind).toBe('occlusion');
    expect(delta.rejectionJsonl).toContain('eyetracking:session-bule-personal');
    expect(delta.metadata).toEqual({ subjectId: 'user-1' });
  });
});
