// @a0n/gnosis/src/flow-obstruction.ts
/**
 * flow-obstruction.ts -- Flow obstruction theory (fluid dynamics of code)
 *
 * A program is a fluid. Control flow is the velocity field. Obstructions are
 * places where flow is impeded -- stagnation points, hydraulic jumps, eddy
 * currents, cavitation bubbles, constrictions, and missing backpressure.
 *
 * Every flow obstruction is a positive-crossing knot. Per
 * knot_positive_deficit_iff_positive_crossings, positive crossings force
 * path collision. Per wallington_produces_unknot, every obstruction admits
 * a finite untangling to the unknot (the obstruction-free program).
 *
 * Mechanized in FlowObstruction.lean (zero sorry):
 *   - stagnation_forces_single_stream: sync in async forces serialization
 *   - hydraulic_jump_adds_crossings: regime mismatch adds crossings
 *   - eddy_never_unknots: recirculation never simplifies (positive beta1)
 *   - unbounded_queue_diverges: no backpressure = unbounded crossing growth
 *   - cavitation_creates_void: unhandled async = information lost to void
 *   - constriction_serializes: lock across yield forces sequential execution
 *   - flow_obstruction_master: obstructions compose monotonically
 *   - obstruction_degrades_rotation: any obstruction worsens rotation makespan
 *   - untangling_recovers_rotation: removing obstructions recovers optimal
 *   - obstructions_never_cancel (anti): two obstructions only compound
 *
 * @module @a0n/gnosis/flow-obstruction
 */

// ============================================================================
// Core Types
// ============================================================================

/** Classification of flow obstructions by fluid dynamics analogy. */
export type ObstructionKind =
  | 'stagnation' // sync call in async context -- flow velocity drops to zero
  | 'hydraulic_jump' // sync/async boundary mismatch -- sudden regime change
  | 'eddy_current' // retry without backoff -- recirculating flow
  | 'backpressure' // unbounded producer -- overflows
  | 'cavitation' // floating promise -- void bubble in flow
  | 'constriction' // lock held across await -- Venturi bottleneck
  | 'sofa'; // dilated payload too complex for manifold -- Gerver bottleneck

/** A flow obstruction -- an annotated algorithmic knot. */
export interface FlowObstruction {
  /** The kind of flow obstruction. */
  readonly kind: ObstructionKind;
  /** Number of crossings (tangles) in the flow graph. Always > 0. */
  readonly crossingNumber: number;
  /** First Betti number -- independent cycles that cannot be untangled. */
  readonly beta1: number;
  /** Number of concurrent paths blocked by this obstruction. */
  readonly blockedPaths: number;
  /** Unknotting number = crossingNumber - beta1.
   *  Number of Reidemeister moves to reach the unknot. */
  readonly unknottingNumber: number;
  /** Lean theorem anchor for this obstruction kind. */
  readonly leanAnchor: string;
}

/** Result of flow obstruction analysis on a function. */
export interface FlowObstructionReport {
  /** Function name (or anonymous). */
  readonly functionName: string;
  /** All obstructions found. */
  readonly obstructions: readonly FlowObstruction[];
  /** Total crossing number (sum of all obstructions). */
  readonly totalCrossings: number;
  /** Effective parallelism after obstructions.
   *  Per obstruction_reduces_parallelism: strictly less than full parallelism. */
  readonly effectiveParallelism: number;
  /** Obstructed makespan overhead: totalCrossings * stageTime.
   *  Per obstruction_degrades_rotation: always positive when obstructions exist. */
  readonly makespanOverhead: number;
}

// ============================================================================
// Factory functions
// ============================================================================

/**
 * Create a stagnation point obstruction.
 * Per stagnation_forces_single_stream: a synchronous call in an async context
 * forces all concurrent paths through a single stream, causing collision.
 */
export function stagnationPoint(asyncPaths: number): FlowObstruction {
  const crossingNumber = Math.max(1, asyncPaths - 1);
  return {
    kind: 'stagnation',
    crossingNumber,
    beta1: 0,
    blockedPaths: asyncPaths,
    unknottingNumber: crossingNumber,
    leanAnchor: 'stagnation_forces_single_stream',
  };
}

/**
 * Create a hydraulic jump obstruction.
 * Per hydraulic_jump_adds_crossings: composing sync and async regimes
 * adds crossings at the boundary beyond the sum of the component knots.
 */
export function hydraulicJump(
  syncCrossings: number,
  asyncCrossings: number,
  boundaryCrossings: number = 1
): FlowObstruction {
  const crossingNumber =
    syncCrossings + asyncCrossings + Math.max(1, boundaryCrossings);
  return {
    kind: 'hydraulic_jump',
    crossingNumber,
    beta1: 0,
    blockedPaths: boundaryCrossings,
    unknottingNumber: crossingNumber,
    leanAnchor: 'hydraulic_jump_adds_crossings',
  };
}

/**
 * Create an eddy current obstruction.
 * Per eddy_never_unknots: recirculation with positive beta1 can never
 * simplify to the unknot -- the cycle structure persists forever.
 */
export function eddyCurrent(loops: number, retries: number): FlowObstruction {
  const crossingNumber = loops + retries;
  return {
    kind: 'eddy_current',
    crossingNumber,
    beta1: loops,
    blockedPaths: retries,
    unknottingNumber: crossingNumber - loops,
    leanAnchor: 'eddy_never_unknots',
  };
}

/**
 * Create a missing backpressure obstruction.
 * Per unbounded_queue_diverges: without backpressure, the queue depth
 * (and thus crossing number) grows without bound.
 */
export function missingBackpressure(queueDepth: number): FlowObstruction {
  return {
    kind: 'backpressure',
    crossingNumber: Math.max(1, queueDepth),
    beta1: 1, // the producer-consumer loop
    blockedPaths: queueDepth,
    unknottingNumber: Math.max(0, queueDepth - 1),
    leanAnchor: 'unbounded_queue_diverges',
  };
}

/**
 * Create a cavitation bubble obstruction.
 * Per cavitation_creates_void: each floating promise is a forked path
 * that was never folded back -- information lost to the void.
 */
export function cavitationBubble(floatingPromises: number): FlowObstruction {
  return {
    kind: 'cavitation',
    crossingNumber: Math.max(1, floatingPromises),
    beta1: 0,
    blockedPaths: floatingPromises,
    unknottingNumber: floatingPromises,
    leanAnchor: 'cavitation_creates_void',
  };
}

/**
 * Create a constriction obstruction.
 * Per constriction_serializes: a lock held across an await point forces
 * all concurrent coroutines through a single stream.
 */
export function constrictionPoint(concurrentPaths: number): FlowObstruction {
  const crossingNumber = Math.max(1, concurrentPaths - 1);
  return {
    kind: 'constriction',
    crossingNumber,
    beta1: 1, // the critical section loop
    blockedPaths: concurrentPaths,
    unknottingNumber: crossingNumber - 1,
    leanAnchor: 'constriction_serializes',
  };
}

/**
 * Create a sofa obstruction.
 * Per sofa_obstruction_leaks_bule: a dilated payload that exceeds the
 * manifold capacity width (Sofa Constant) creates a bottleneck that 
 * requires extreme Reidemeister maneuvering to untangle.
 * 
 * Crossing number is the area deficit multiplied by normalized complexity.
 */
export function sofaObstruction(
  area: number,
  width: number,
  constant: number = 2.219531669
): FlowObstruction {
  const ratio = area / Math.max(0.1, width);
  const crossingNumber = ratio > constant 
    ? Math.max(1, Math.floor((ratio - constant) * 100)) 
    : 0;

  return {
    kind: 'sofa',
    crossingNumber,
    beta1: ratio > constant ? 18 : 0, // Gerver 18-arc complexity
    blockedPaths: ratio > constant ? 1 : 0,
    unknottingNumber: crossingNumber,
    leanAnchor: 'sofa_obstruction_leaks_bule',
  };
}

// ============================================================================
// Composition -- per flow_obstruction_master
// ============================================================================

/**
 * Compose two obstructions. Per flow_obstruction_master: the composed
 * crossing number strictly exceeds either component. Per obstructions_never_cancel:
 * they never simplify each other, only compound.
 */
export function composeObstructions(
  a: FlowObstruction,
  b: FlowObstruction
): FlowObstruction {
  return {
    kind: a.kind, // primary obstruction kind
    crossingNumber: a.crossingNumber + b.crossingNumber,
    beta1: a.beta1 + b.beta1,
    blockedPaths: a.blockedPaths + b.blockedPaths,
    unknottingNumber: a.unknottingNumber + b.unknottingNumber,
    leanAnchor: 'flow_obstruction_master',
  };
}

// ============================================================================
// Scheduling cost -- per obstruction_degrades_rotation
// ============================================================================

/**
 * Compute the scheduling overhead of an obstruction.
 * Per obstruction_cost_exact: cost = crossingNumber * stageTime.
 * Per obstruction_degrades_rotation: this is always positive.
 */
export function obstructionCost(
  obs: FlowObstruction,
  stageTime: number = 1
): number {
  return obs.crossingNumber * stageTime;
}

/**
 * Compute effective parallelism after obstructions.
 * Per obstruction_reduces_parallelism: strictly less than full parallelism
 * when any obstruction blocks paths.
 */
export function effectiveParallelism(
  totalPaths: number,
  obs: FlowObstruction
): number {
  return Math.max(1, totalPaths - obs.blockedPaths);
}

// ============================================================================
// Report builder
// ============================================================================

/**
 * Build a flow obstruction report for a function.
 * Per composed_obstruction_cost: total cost is the sum of individual costs.
 */
export function buildReport(
  functionName: string,
  obstructions: FlowObstruction[],
  totalPaths: number = 4
): FlowObstructionReport {
  const totalCrossings = obstructions.reduce(
    (sum, o) => sum + o.crossingNumber,
    0
  );
  const totalBlocked = obstructions.reduce((sum, o) => sum + o.blockedPaths, 0);
  return {
    functionName,
    obstructions,
    totalCrossings,
    effectiveParallelism: Math.max(1, totalPaths - totalBlocked),
    makespanOverhead: totalCrossings,
  };
}

// ============================================================================
// Performance obstructions -- per PerformanceObstruction.lean
// ============================================================================

/** Classification of performance obstructions. */
export type PerfObstructionKind =
  | 'quadratic_accumulate' // string += in loop: O(n^2) copy
  | 'unbounded_regex' // nested quantifiers: exponential backtracking
  | 'alloc_in_hot_loop' // allocation per iteration: N * allocCost
  | 'serial_await' // sequential await in loop: N * latency
  | 'map_lookup_recompute' // redundant identical lookups: (k-1) * cost wasted
  | 'cold_sort_scan' // sort then take first: n*log(n) for O(n) work
  | 'nested_loop_search' // inner loop linear search: n*m vs n+m
  | 'sync_copy_oversized'; // serialize + deserialize round-trip: 2*N crossings

/** A performance obstruction with actual and optimal costs. */
export interface PerformanceObstruction {
  readonly kind: PerfObstructionKind;
  /** Actual crossing cost of the anti-pattern. */
  readonly actualCost: number;
  /** Optimal crossing cost of the fix. */
  readonly optimalCost: number;
  /** Wasted crossings = actualCost - optimalCost. Always > 0. */
  readonly perfGap: number;
  /** Lean theorem anchor. */
  readonly leanAnchor: string;
}

/** Per quadratic_accumulate_cost: n copy-appends = n*(n+1)/2. */
export function quadraticAccumulate(
  iterations: number
): PerformanceObstruction {
  const actual = Math.floor((iterations * (iterations + 1)) / 2);
  return {
    kind: 'quadratic_accumulate',
    actualCost: actual,
    optimalCost: iterations,
    perfGap: actual - iterations,
    leanAnchor: 'quadratic_accumulate_cost',
  };
}

/** Per regex_backtrack_exponential: depth d → 2^d paths. */
export function unboundedRegex(
  quantifierDepth: number
): PerformanceObstruction {
  const actual = Math.pow(2, quantifierDepth);
  return {
    kind: 'unbounded_regex',
    actualCost: actual,
    optimalCost: 1,
    perfGap: actual - 1,
    leanAnchor: 'regex_backtrack_exponential',
  };
}

/** Per allocation_multiplies_crossings: N * allocCost. */
export function allocInHotLoop(
  iterations: number,
  allocCost: number = 1
): PerformanceObstruction {
  const actual = iterations * allocCost;
  return {
    kind: 'alloc_in_hot_loop',
    actualCost: actual,
    optimalCost: allocCost,
    perfGap: (iterations - 1) * allocCost,
    leanAnchor: 'allocation_multiplies_crossings',
  };
}

/** Per serial_await_speedup: speedup = N. */
export function serialAwait(
  iterations: number,
  latency: number = 1
): PerformanceObstruction {
  return {
    kind: 'serial_await',
    actualCost: iterations * latency,
    optimalCost: latency,
    perfGap: (iterations - 1) * latency,
    leanAnchor: 'serial_await_speedup',
  };
}

/** Per redundant_lookup_cost: (k-1) * lookupCost wasted. */
export function mapLookupRecompute(
  lookups: number,
  lookupCost: number = 1
): PerformanceObstruction {
  return {
    kind: 'map_lookup_recompute',
    actualCost: lookups * lookupCost,
    optimalCost: lookupCost,
    perfGap: (lookups - 1) * lookupCost,
    leanAnchor: 'redundant_lookup_cost',
  };
}

/** Per sort_then_first_wasteful: gap = n * (logBound - 1). */
export function coldSortScan(
  n: number,
  logBound: number = 2
): PerformanceObstruction {
  return {
    kind: 'cold_sort_scan',
    actualCost: n * logBound,
    optimalCost: n,
    perfGap: n * (logBound - 1),
    leanAnchor: 'sort_then_first_wasteful',
  };
}

/** Per nested_search_quadratic: n*m vs n+m. */
export function nestedLoopSearch(
  outer: number,
  inner: number
): PerformanceObstruction {
  return {
    kind: 'nested_loop_search',
    actualCost: outer * inner,
    optimalCost: outer + inner,
    perfGap: outer * inner - (outer + inner),
    leanAnchor: 'nested_search_quadratic',
  };
}

/** Per copy_doubles_crossings: 2 * structSize. */
export function syncCopyOversized(structSize: number): PerformanceObstruction {
  return {
    kind: 'sync_copy_oversized',
    actualCost: 2 * structSize,
    optimalCost: structSize,
    perfGap: structSize,
    leanAnchor: 'copy_doubles_crossings',
  };
}
