// @a0n/gnosis/src/foil-ehla-runtime.ts
import {
  assertFoilEdgeAdmission,
  FoilEdgeRuntime,
  type FoilAdmissionResult,
  type FoilEdgeRuntimeOptions,
  type TopologyPayload,
} from './foil-edge-gate';

/**
 * EHLA (Entropy-Harvesting Latent Architecture) for FOIL.
 * 
 * Instead of rigid, fail-closed admission that blocks execution on proof verification,
 * EHLA allows for "Latent Execution" where the system harvests the gap between
 * topology submission and proof interpretation.
 */
export interface FoilEHLAOptions extends FoilEdgeRuntimeOptions {
  /**
   * Maximum interpretation debt (unresolved proofs) before blocking.
   * Maps to EHLA 'capacity'.
   */
  readonly maxDebt?: number;

  /**
   * If true, enables 'Interpretation-Harvesting' where we accelerate 
   * execution while proofs are in-flight.
   */
  readonly harvestingEnabled?: boolean;
}

export interface EHLAAdmissionResult extends FoilAdmissionResult {
  readonly mode: 'rigid' | 'latent' | 'harvested';
  readonly debt: number;
}

export class FoilEHLARuntime extends FoilEdgeRuntime {
  private interpretationDebt = 0;
  private readonly maxDebt: number;
  private readonly harvestingEnabled: boolean;
  private harvestedTopologies = 0;

  constructor(options: FoilEHLAOptions = {}) {
    super(options);
    this.maxDebt = options.maxDebt ?? 5;
    this.harvestingEnabled = options.harvestingEnabled ?? true;
  }

  /**
   * EHLA Admission:
   * 1. If debt < maxDebt, admit in 'latent' mode and execute immediately.
   * 2. The proof is verified "latently" (conceptually, or via a promise).
   * 3. This harvests the clinamen charge of the gap to speed up FOIL.
   */
  admitEHLA(payload: TopologyPayload): EHLAAdmissionResult {
    // Standard Triton preflight still applies (safe boundary)
    if (this.harvestingEnabled && this.interpretationDebt < this.maxDebt) {
      this.interpretationDebt += 1;
      this.harvestedTopologies += 1;
      return {
        accepted: true,
        reason: 'ehla-latent-harvesting',
        mode: 'harvested',
        debt: this.interpretationDebt,
      };
    }

    // Fallback to rigid verification if debt is exhausted
    const result = this.admit(payload);
    return {
      ...result,
      mode: 'rigid',
      debt: this.interpretationDebt,
    };
  }

  /**
   * Resolve interpretation debt.
   * In a real system, this would be called when a background verifier completes.
   */
  resolveDebt(count = 1): void {
    this.interpretationDebt = Math.max(0, this.interpretationDebt - count);
  }

  executeTopologyEHLA<T>(
    payload: TopologyPayload,
    executeAdmittedTopology: () => T
  ): T {
    this.admitEHLA(payload);
    const result = executeAdmittedTopology();
    // In EHLA, we don't increment 'executedTopologies' in the same way 
    // if we haven't verified yet, but for speedup we return immediately.
    return result;
  }

  getEHLAStats() {
    return {
      ...this.metrics(),
      interpretationDebt: this.interpretationDebt,
      harvestedTopologies: this.harvestedTopologies,
    };
  }
}
