// @a0n/gnosis/src/formal-ledger.ts
import { INV_PHI } from './math';
import {
  createBraidPacketCertificate,
  createExtendedAeonTransitionCertificate,
  createGnosticMetrologyProfile,
  createGnosticLensGeometryCertificate,
  createGnosisUringInterferePortCertificate,
  createInterferenceStrideCertificate,
  createPressureGradientCertificate,
  createPressureCertifiedExecutionPathCertificate,
  createPressureWeightedInterferenceStrideCertificate,
  createStatisticalMechanicsAccounting,
  type BraidPacketCertificate,
  type ExtendedAeonTransitionCertificate,
  type GnosticMetrologyProfile,
  type GnosticLensGeometryCertificate,
  type GnosisUringInterferePortCertificate,
  type InterferenceStrideCertificate,
  type PressureGradientCertificate,
  type PressureCertifiedExecutionPathCertificate,
  type PressureWeightedInterferenceStrideCertificate,
  type StatisticalMechanicsAccountingCertificate,
} from './mesh-policy';

export type {
  BraidPacketCertificate,
  CalibratedFluidicPressureBridge,
  ExtendedAeonTransitionCertificate,
  FluidicRoutingAdmissibilityCertificate,
  GnosticMetrologyProfile,
  GnosticLensGeometryCertificate,
  GnosisUringInterferePortCertificate,
  GnosticRuntimeAllocationPlan,
  InterferenceStrideCertificate,
  PressureGradientCertificate,
  PressureCertifiedExecutionPathCertificate,
  PressureWeightedInterferenceStrideCertificate,
  StatisticalMechanicsAccountingCertificate,
} from './mesh-policy';

export {
  createBraidPacketCertificate,
  createExtendedAeonTransitionCertificate,
  createFluidicRoutingAdmissibilityCertificate,
  createGnosisUringInterferePortCertificate,
  createGnosticLensGeometryCertificate,
  createInterferenceStrideCertificate,
  createPressureWeightedInterferenceStrideCertificate,
  createPressureCertifiedExecutionPathCertificate,
} from './mesh-policy';

export const FORMAL_LEDGER_FAMILIES = [
  'voidWalking',
  'negotiation',
  'community',
  'metacognition',
  'recovery',
  'probability',
] as const;

export const FORMAL_LEDGER_PRIMITIVES = [
  'fork',
  'race',
  'fold',
  'vent',
  'interfere',
] as const;

export type FormalLedgerFamilyId = (typeof FORMAL_LEDGER_FAMILIES)[number];
export type FormalLedgerPrimitiveId = (typeof FORMAL_LEDGER_PRIMITIVES)[number];
export type FormalLedgerSeverity = 'low' | 'moderate' | 'high';
export type FormalLedgerAudience = 'wellness' | 'research';
export type FormalLedgerVisualizationKind = 'radar' | 'torus' | 'trace';
export type FormalLedgerRecommendationPriority =
  | 'support'
  | 'stabilize'
  | 'urgent';

export interface FormalLedgerServiceDeclaration {
  readonly serviceId: string;
  readonly consume?: readonly FormalLedgerFamilyId[];
  readonly emit?: readonly FormalLedgerFamilyId[];
  readonly optedOut?: boolean;
  readonly optOutReason?: string;
}

export interface FormalLedgerPersonalityInput {
  readonly [key: string]: unknown;
  readonly try_?: number;
  readonly choose?: number;
  readonly commit?: number;
  readonly letGo?: number;
  readonly learn?: number;
  readonly source?: string;
}

export interface FormalLedgerNegotiationInput {
  readonly [key: string]: unknown;
  readonly assertiveness?: number;
  readonly cooperation?: number;
  readonly batnaStrength?: number;
  readonly watnaRisk?: number;
  readonly settlementReadiness?: number;
  readonly trustReserve?: number;
  readonly source?: string;
}

export interface FormalLedgerCommunityInput {
  readonly [key: string]: unknown;
  readonly empathy?: number;
  readonly belonging?: number;
  readonly dominanceRisk?: number;
  readonly mergePressure?: number;
  readonly reciprocity?: number;
  readonly source?: string;
}

export interface FormalLedgerMetacognitionInput {
  readonly [key: string]: unknown;
  readonly c0Confidence?: number;
  readonly c1Calibration?: number;
  readonly c2Coherence?: number;
  readonly c3BiasAwareness?: number;
  readonly source?: string;
}

export interface FormalLedgerRecoveryInput {
  readonly [key: string]: unknown;
  readonly stressLoad?: number;
  readonly sleepDebt?: number;
  readonly restoration?: number;
  readonly forgiveness?: number;
  readonly breathRegulation?: number;
  readonly rumination?: number;
  readonly source?: string;
}

export interface FormalLedgerProbabilityInput {
  readonly [key: string]: unknown;
  readonly hypothesisCount?: number;
  readonly topPosterior?: number;
  readonly runnerUpPosterior?: number;
  readonly evidenceTraceability?: number;
  readonly unsupportedClaimRatio?: number;
  readonly validationStrength?: number;
  readonly updateStrength?: number;
  readonly convergenceRate?: number;
  readonly source?: string;
}

export interface FormalLedgerProbabilityAnalysisEntry {
  readonly priorProbability: number;
  readonly posteriorProbability: number;
}

export interface FormalLedgerProbabilityAnalysisAudit {
  readonly traceabilityCoverage: number;
  readonly unsupportedClaims: readonly string[];
}

export interface FormalLedgerProbabilityAnalysisValidationMetric {
  readonly value: number;
}

export interface FormalLedgerProbabilityAnalysisValidation {
  readonly skipped: boolean;
  readonly holdout?: {
    readonly correlation: FormalLedgerProbabilityAnalysisValidationMetric;
    readonly meanAbsoluteError: FormalLedgerProbabilityAnalysisValidationMetric;
  };
  readonly negativeControl?: {
    readonly degradation: number;
  };
}

export interface FormalLedgerProbabilityAnalysis {
  readonly hypothesisRanking: readonly FormalLedgerProbabilityAnalysisEntry[];
  readonly evidenceAudit: FormalLedgerProbabilityAnalysisAudit;
  readonly validation: FormalLedgerProbabilityAnalysisValidation;
  readonly diagnostics: {
    readonly convergence: {
      readonly observedRate: number;
    };
  };
}

export type FormalLedgerNegotiationModeId =
  | 'competing'
  | 'collaborating'
  | 'compromising'
  | 'avoiding'
  | 'accommodating';

export type FormalLedgerNegotiationPressureBias = 'soften' | 'hold' | 'push';

export interface FormalLedgerNegotiationAnalysisRound {
  readonly alignment: number;
}

export interface FormalLedgerNegotiationAnalysisTrajectoryStep {
  readonly alignment: number;
}

export interface FormalLedgerNegotiationAnalysisStrategicPressure {
  readonly alignment?: number;
  readonly volatility?: number;
  readonly escalationRisk?: number;
  readonly settlementPressure?: number;
}

export interface FormalLedgerNegotiationAnalysisSettlementCandidate {
  readonly confidence?: number;
}

export interface FormalLedgerNegotiationAnalysis {
  readonly tkiLikeSummary: {
    readonly primaryMode: FormalLedgerNegotiationModeId;
    readonly secondaryMode: FormalLedgerNegotiationModeId;
  };
  readonly bazaar: {
    readonly rounds: readonly FormalLedgerNegotiationAnalysisRound[];
    readonly branchOutcomes?: readonly unknown[];
    readonly settled?: boolean;
    readonly trajectory?: readonly FormalLedgerNegotiationAnalysisTrajectoryStep[];
    readonly strategicPressure?: FormalLedgerNegotiationAnalysisStrategicPressure;
  };
  readonly neutral: {
    readonly convergenceStatus?: string;
    readonly settlementCandidates: readonly FormalLedgerNegotiationAnalysisSettlementCandidate[];
  };
  readonly neutralSettlementBias: {
    readonly pressureBias?: FormalLedgerNegotiationPressureBias;
  };
}

export type FormalLedgerCommunicationAlertSeverity =
  | 'low'
  | 'medium'
  | 'high'
  | 'critical';

export interface FormalLedgerCommunicationAlert {
  readonly severity: FormalLedgerCommunicationAlertSeverity;
  readonly twinTookAction?: boolean;
}

export interface FormalLedgerCommunicationAnalysis {
  readonly topology: {
    readonly participantCount: number;
    readonly outputStreams?: number;
    readonly agreementRatio?: number;
    readonly sharedContextUnits?: number;
  };
  readonly signals: {
    readonly empathySignal?: number;
    readonly reciprocitySignal?: number;
    readonly dominanceRisk?: number;
    readonly mergePressure?: number;
    readonly clarificationNeed?: number;
    readonly perspectiveBreadth?: number;
    readonly authenticity?: number;
  };
  readonly safety?: {
    readonly entropyLevel?: number;
    readonly alerts?: readonly FormalLedgerCommunicationAlert[];
    readonly stressLoad?: number;
    readonly sleepDebt?: number;
    readonly restorationCapacity?: number;
    readonly breathRegulation?: number;
    readonly ruminationRisk?: number;
  };
  readonly source?: string;
}

export interface FormalLedgerMetacognitionAnalysisLayer {
  readonly rawConfidence?: number;
  readonly calibratedConfidence?: number;
  readonly dataQuality?: number;
  readonly credibleInterval?: readonly [number, number];
}

export interface FormalLedgerMetacognitionAnalysis {
  readonly calibration: {
    readonly layers?: readonly FormalLedgerMetacognitionAnalysisLayer[];
    readonly overallConfidence?: number;
  };
  readonly coherence: {
    readonly overallCoherence?: number;
    readonly coherenceImprovement?: number;
    readonly remainingTensionCount?: number;
    readonly resolvedContradictionCount?: number;
    readonly layerAlignmentScores?: readonly number[];
  };
  readonly bias: {
    readonly detectedCount?: number;
    readonly correctedCount?: number;
    readonly confidenceInCorrections?: number;
    readonly totalDebiasedDelta?: number;
    readonly affectedLayerBreadth?: number;
  };
  readonly quality?: {
    readonly overallQuality?: number;
  };
  readonly source?: string;
}

export interface FormalLedgerRecoveryAnalysis {
  readonly signals: {
    readonly stressLoad?: number;
    readonly sleepDebt?: number;
    readonly recoveryTimeMinutes?: number;
    readonly burnoutRisk?: number;
    readonly restorationCapacity?: number;
    readonly forgiveness?: number;
    readonly breathRegulation?: number;
    readonly ruminationRisk?: number;
    readonly confidence?: number;
  };
  readonly markers?: {
    readonly indicators?: readonly string[];
    readonly recoveryFactors?: readonly string[];
  };
  readonly source?: string;
}

export interface FormalLedgerJusticeAnalysis {
  readonly evidence: {
    readonly totalThreads?: number;
    readonly coveredThreads?: number;
    readonly contestedThreads?: number;
    readonly uncoveredThreads?: number;
    readonly evidenceDeficit?: number;
    readonly averageConfidence?: number;
    readonly sufficient?: boolean;
  };
  readonly adjudication?: {
    readonly status?: string;
    readonly remedyCandidateCount?: number;
    readonly rankedRemedyCount?: number;
    readonly affectedCommunityCount?: number;
    readonly communitySignalCount?: number;
    readonly ratificationRequired?: boolean;
    readonly enforcementActive?: boolean;
    readonly appealAvailable?: boolean;
  };
  readonly process?: {
    readonly criticalIssueCount?: number;
    readonly injusticeTriggerCount?: number;
    readonly recommendedActionCount?: number;
    readonly summaryClarity?: number;
  };
  readonly source?: string;
}

export interface FormalLedgerGovernanceAnalysisRanking {
  readonly candidateId: string;
  readonly approvals?: number;
  readonly rejections?: number;
  readonly representedCommunities?: number;
  readonly uncoveredCommunities?: number;
  readonly deficit?: number;
}

export interface FormalLedgerGovernanceAnalysis {
  readonly totalCommunities?: number;
  readonly rankings: readonly FormalLedgerGovernanceAnalysisRanking[];
  readonly winner?: FormalLedgerGovernanceAnalysisRanking | null;
  readonly process?: {
    readonly consensusLevel?: number;
    readonly dissentCount?: number;
    readonly delegationAvailable?: boolean;
    readonly emotionalCostIndex?: string;
    readonly condorcetWinner?: boolean;
  };
  readonly source?: string;
}

export interface FormalLedgerTrustAnalysis {
  readonly authority?: {
    readonly verifiedActor?: boolean;
    readonly delegationDepth?: number;
    readonly capabilityCoverage?: number;
    readonly overridePressure?: number;
    readonly holdCount?: number;
    readonly activeHoldRatio?: number;
    readonly quarantineRisk?: number;
    readonly forgeHoldRisk?: number;
  };
  readonly signals: {
    readonly manipulationLoad?: number;
    readonly factualCoverage?: number;
    readonly impliedPressure?: number;
    readonly authenticityGap?: number;
    readonly requestClarity?: number;
    readonly evidenceBreadth?: number;
  };
  readonly source?: string;
}

export type FormalLedgerBehaviorValence = 'positive' | 'negative' | 'neutral';
export type FormalLedgerBehaviorRiskLevel = 'low' | 'moderate' | 'high';

export interface FormalLedgerBehaviorAnalysisDetection {
  readonly confidence?: number;
  readonly valence?: FormalLedgerBehaviorValence;
  readonly matchedTermCount?: number;
  readonly hasIntervention?: boolean;
}

export interface FormalLedgerBehaviorAnalysis {
  readonly detections: readonly FormalLedgerBehaviorAnalysisDetection[];
  readonly summary?: {
    readonly dominantValence?: FormalLedgerBehaviorValence | 'mixed';
    readonly riskLevel?: FormalLedgerBehaviorRiskLevel;
    readonly actionableCount?: number;
  };
  readonly recommendationCount?: number;
  readonly source?: string;
}

export interface FormalLedgerContext {
  readonly declaration?: FormalLedgerServiceDeclaration;
  readonly generatedAt?: string;
  readonly personality?: FormalLedgerPersonalityInput;
  readonly negotiation?: FormalLedgerNegotiationInput;
  readonly community?: FormalLedgerCommunityInput;
  readonly metacognition?: FormalLedgerMetacognitionInput;
  readonly recovery?: FormalLedgerRecoveryInput;
  readonly probability?: FormalLedgerProbabilityInput;
}

export interface FormalLedgerPrimitiveState {
  readonly id: FormalLedgerPrimitiveId;
  readonly title: string;
  readonly strength: number;
  readonly deficit: number;
  readonly severity: FormalLedgerSeverity;
  readonly summary: string;
}

export interface FormalLedgerUnmappedSegment {
  readonly id: string;
  readonly label: string;
  readonly value: number;
  readonly summary: string;
}

export interface FormalLedgerDiagnostic {
  readonly id: string;
  readonly familyId: FormalLedgerFamilyId;
  readonly title: string;
  readonly summary: string;
  readonly detail?: string;
  readonly severity: FormalLedgerSeverity;
  readonly primitiveId?: FormalLedgerPrimitiveId;
  readonly value?: number;
  readonly sourceRefs: readonly string[];
}

export interface FormalLedgerNarrative {
  readonly id: string;
  readonly familyId: FormalLedgerFamilyId;
  readonly audience: FormalLedgerAudience;
  readonly title: string;
  readonly summary: string;
  readonly detail?: string;
}

export interface FormalLedgerRecommendation {
  readonly id: string;
  readonly familyId: FormalLedgerFamilyId;
  readonly title: string;
  readonly rationale: string;
  readonly actions: readonly string[];
  readonly modalities: readonly string[];
  readonly priority: FormalLedgerRecommendationPriority;
}

export interface FormalLedgerVisualization {
  readonly id: string;
  readonly familyId: FormalLedgerFamilyId;
  readonly kind: FormalLedgerVisualizationKind;
  readonly title: string;
  readonly summary: string;
  readonly primitives: readonly FormalLedgerPrimitiveState[];
  readonly highlightedPrimitiveId?: FormalLedgerPrimitiveId;
  readonly metrics: {
    readonly balanceScore: number;
    readonly activeDeficit: number;
    readonly referenceAttractor: number;
  };
}

export interface FormalLedgerFamilyReport {
  readonly familyId: FormalLedgerFamilyId;
  readonly title: string;
  readonly active: boolean;
  readonly summary: string;
  readonly sources: readonly string[];
  readonly primitives: readonly FormalLedgerPrimitiveState[];
  readonly diagnostics: readonly FormalLedgerDiagnostic[];
  readonly narratives: readonly FormalLedgerNarrative[];
  readonly recommendations: readonly FormalLedgerRecommendation[];
  readonly visualization: FormalLedgerVisualization;
  readonly unmappedSegments: readonly FormalLedgerUnmappedSegment[];
  readonly metadata: Readonly<Record<string, string | number | boolean>>;
}

export interface FormalLedgerStatisticalMechanicsAccounting {
  readonly source: 'GnosisFormalLedgerStatisticalMechanicsAccounting';
  readonly certificate: StatisticalMechanicsAccountingCertificate;
  readonly metrology: GnosticMetrologyProfile;
  readonly pressureGradient: PressureGradientCertificate;
  readonly braidPacket: BraidPacketCertificate;
  readonly extendedAeon: ExtendedAeonTransitionCertificate;
  readonly interferePort: GnosisUringInterferePortCertificate;
  readonly interferenceStride: InterferenceStrideCertificate;
  readonly pressureWeightedInterference: PressureWeightedInterferenceStrideCertificate;
  readonly gnosticLensGeometry: GnosticLensGeometryCertificate;
  readonly summary: string;
  readonly sourceRefs: readonly string[];
}

export interface FormalLedgerBundle {
  readonly generatedAt: string;
  readonly declaration?: FormalLedgerServiceDeclaration;
  readonly familyOrder: readonly FormalLedgerFamilyId[];
  readonly activeFamilies: readonly FormalLedgerFamilyId[];
  readonly primitives: readonly FormalLedgerPrimitiveState[];
  readonly reports: Readonly<
    Record<FormalLedgerFamilyId, FormalLedgerFamilyReport>
  >;
  readonly diagnostics: readonly FormalLedgerDiagnostic[];
  readonly narratives: readonly FormalLedgerNarrative[];
  readonly recommendations: readonly FormalLedgerRecommendation[];
  readonly visualizations: readonly FormalLedgerVisualization[];
  readonly statisticalMechanics: FormalLedgerStatisticalMechanicsAccounting;
  readonly sourceMenu: readonly string[];
  readonly compliance: {
    readonly declaredFamilies: readonly FormalLedgerFamilyId[];
    readonly undeclaredFamilies: readonly FormalLedgerFamilyId[];
    readonly optedOut: boolean;
    readonly status: 'declared' | 'undeclared' | 'opted_out';
  };
}

interface FamilyCatalogEntry {
  readonly title: string;
  readonly sources: readonly string[];
  readonly kind: FormalLedgerVisualizationKind;
}

function buildFormalLedgerContext(
  declaration: FormalLedgerServiceDeclaration | undefined,
  extra: Omit<FormalLedgerContext, 'declaration'> = {}
): FormalLedgerContext {
  return declaration === undefined
    ? { ...extra }
    : { declaration, ...extra };
}

// interface _FamilyShape {
//   readonly primitives: Readonly<Record<FormalLedgerPrimitiveId, number>>;
//   readonly summary: string;
//   readonly sources: readonly string[];
//   readonly kind: FormalLedgerVisualizationKind;
//   readonly diagnostics: readonly FormalLedgerDiagnostic[];
//   readonly narratives: readonly FormalLedgerNarrative[];
//   readonly recommendations: readonly FormalLedgerRecommendation[];
//   readonly unmappedSegments: readonly FormalLedgerUnmappedSegment[];
//   readonly metadata: Readonly<Record<string, string | number | boolean>>;
// }

const PRIMITIVE_TITLES: Readonly<Record<FormalLedgerPrimitiveId, string>> = {
  fork: 'Fork',
  race: 'Race',
  fold: 'Fold',
  vent: 'Vent',
  interfere: 'Interfere',
};

const NEGOTIATION_MODE_SCORES: Readonly<
  Record<
    FormalLedgerNegotiationModeId,
    { readonly assertiveness: number; readonly cooperation: number }
  >
> = {
  competing: { assertiveness: 0.88, cooperation: 0.22 },
  collaborating: { assertiveness: 0.74, cooperation: 0.84 },
  compromising: { assertiveness: 0.62, cooperation: 0.63 },
  avoiding: { assertiveness: 0.18, cooperation: 0.28 },
  accommodating: { assertiveness: 0.22, cooperation: 0.82 },
};

const NEGOTIATION_PRESSURE_BIAS_RISK: Readonly<
  Record<FormalLedgerNegotiationPressureBias, number>
> = {
  soften: 0.28,
  hold: 0.52,
  push: 0.82,
};

const FAMILY_CATALOG: Readonly<
  Record<FormalLedgerFamilyId, FamilyCatalogEntry>
> = {
  voidWalking: {
    title: 'Void Walking',
    kind: 'radar',
    sources: [
      'open-source/gnosis/THEOREM_LEDGER.md',
      '@a0n/buleyean-rl/personality.ts',
      '@a0n/maybe',
    ],
  },
  negotiation: {
    title: 'Negotiation',
    kind: 'trace',
    sources: [
      'open-source/gnosis/THEOREM_LEDGER.md',
      '@a0n/aeon-bazaar',
      '@a0n/aeon-neutral',
    ],
  },
  community: {
    title: 'Community',
    kind: 'torus',
    sources: [
      'open-source/gnosis/THEOREM_LEDGER.md',
      'CommunityDominance.lean',
      'SkyrmsNadirBule.lean',
    ],
  },
  metacognition: {
    title: 'Metacognition',
    kind: 'trace',
    sources: [
      'open-source/gnosis/THEOREM_LEDGER.md',
      'MetacognitiveDaisyChain.lean',
    ],
  },
  recovery: {
    title: 'Recovery',
    kind: 'radar',
    sources: [
      'open-source/gnosis/THEOREM_LEDGER.md',
      'SleepDebt.lean',
      'SleepDebtSchedule.lean',
    ],
  },
  probability: {
    title: 'Probability',
    kind: 'trace',
    sources: [
      'open-source/gnosis/THEOREM_LEDGER.md',
      '@a0n/maybe',
      'open-source/maybe/src/layers.ts',
    ],
  },
};

/**
 * Handles the gnosis derive Formal Ledger Bundle workflow.
 */
export function deriveFormalLedgerBundle(
  context: FormalLedgerContext
): FormalLedgerBundle {
  const reports = {
    voidWalking: createVoidWalkingReport(context.personality),
    negotiation: createNegotiationReport(context.negotiation),
    community: createCommunityReport(context.community),
    metacognition: createMetacognitionReport(context.metacognition),
    recovery: createRecoveryReport(context.recovery),
    probability: createProbabilityReport(context.probability),
  } satisfies Record<FormalLedgerFamilyId, FormalLedgerFamilyReport>;

  const activeFamilies = FORMAL_LEDGER_FAMILIES.filter(
    (familyId) => reports[familyId].active
  );
  const bundlePrimitives = aggregateBundlePrimitives(reports, activeFamilies);
  const diagnostics = activeFamilies.flatMap(
    (familyId) => reports[familyId].diagnostics
  );
  const narratives = activeFamilies.flatMap(
    (familyId) => reports[familyId].narratives
  );
  const recommendations = activeFamilies.flatMap(
    (familyId) => reports[familyId].recommendations
  );
  const visualizations = activeFamilies.map(
    (familyId) => reports[familyId].visualization
  );
  const compliance = createCompliance(context.declaration, activeFamilies);
  const complianceDiagnostic = createComplianceDiagnostic(compliance);
  const generatedAt = context.generatedAt ?? new Date().toISOString();
  const statisticalMechanics =
    createFormalLedgerStatisticalMechanicsAccounting();
  const sourceMenu = Array.from(
    new Set([
      ...FORMAL_LEDGER_FAMILIES.flatMap(
        (familyId) => reports[familyId].sources
      ),
      ...statisticalMechanics.sourceRefs,
    ])
  );

  return {
    generatedAt,
    ...(context.declaration !== undefined
      ? { declaration: context.declaration }
      : {}),
    familyOrder: FORMAL_LEDGER_FAMILIES,
    activeFamilies,
    primitives: bundlePrimitives,
    reports,
    diagnostics:
      complianceDiagnostic == null
        ? diagnostics
        : [...diagnostics, complianceDiagnostic],
    narratives,
    recommendations,
    visualizations,
    statisticalMechanics,
    sourceMenu,
    compliance,
  };
}

/**
 * Creates the Formal Ledger Statistical Mechanics Accounting.
 */
export function createFormalLedgerStatisticalMechanicsAccounting(): FormalLedgerStatisticalMechanicsAccounting {
  const certificate = createStatisticalMechanicsAccounting();
  if (certificate === null) {
    throw new Error(
      'Exact statistical mechanics accounting certificate could not be derived from the formal mesh-policy surface.'
    );
  }
  const metrology = createGnosticMetrologyProfile(certificate);
  if (metrology === null) {
    throw new Error(
      'Exact gnostic metrology profile could not be derived from the statistical mechanics accounting certificate.'
    );
  }
  const pressureGradient = createPressureGradientCertificate(metrology);
  if (pressureGradient === null) {
    throw new Error(
      'Exact pressure-gradient certificate could not be derived from the gnostic metrology profile.'
    );
  }
  const braidPacket = createBraidPacketCertificate(0, certificate);
  if (braidPacket === null) {
    throw new Error(
      'Exact braid-packet certificate could not be derived from the statistical mechanics accounting certificate.'
    );
  }
  const extendedAeon = createExtendedAeonTransitionCertificate(braidPacket);
  if (extendedAeon === null) {
    throw new Error(
      'Exact extended-aeon certificate could not be derived from the braid packet certificate.'
    );
  }
  const interferePort = createGnosisUringInterferePortCertificate(extendedAeon);
  if (interferePort === null) {
    throw new Error(
      'Exact gnosis-uring interfere-port certificate could not be derived from the extended-aeon certificate.'
    );
  }
  const interferenceStride = createInterferenceStrideCertificate(braidPacket);
  if (interferenceStride === null) {
    throw new Error(
      'Exact interference-stride certificate could not be derived from the braid packet certificate.'
    );
  }
  const pressureWeightedInterference =
    createPressureWeightedInterferenceStrideCertificate(interferenceStride);
  if (pressureWeightedInterference === null) {
    throw new Error(
      'Exact pressure-weighted interference certificate could not be derived from the interference-stride certificate.'
    );
  }
  const gnosticLensGeometry = createGnosticLensGeometryCertificate(braidPacket);
  if (gnosticLensGeometry === null) {
    throw new Error(
      'Exact gnostic-lens geometry certificate could not be derived from the braid packet certificate.'
    );
  }

  return {
    source: 'GnosisFormalLedgerStatisticalMechanicsAccounting',
    certificate,
    metrology,
    pressureGradient,
    braidPacket,
    extendedAeon,
    interferePort,
    interferenceStride,
    pressureWeightedInterference,
    gnosticLensGeometry,
    summary:
      'Exact 10/30/90 saturation accounting over the 12-row Aeon budget: 1560 gG total mass, 96 SPa total pressure, with the 43 interfere port certified as a terminal gnosis-uring handoff, pressure-weighted stride anchor, and finite gnostic-lens geometry certificate.',
    sourceRefs: [
      'lean/Lean/ForkRaceFoldTheorems/GamutBraid.lean',
      'lean/Lean/ForkRaceFoldTheorems/StatisticalMechanicsAccounting.lean',
      'src/mesh-policy.ts',
      'open-source/gnosis/FORMAL_LEDGER.md#statistical-mechanics-accounting-information-beach',
    ],
  };
}

/**
 * Creates the Probability Formal Ledger Input.
 */
export function createProbabilityFormalLedgerInput(
  analysis: FormalLedgerProbabilityAnalysis
): FormalLedgerProbabilityInput {
  const topPosterior =
    analysis.hypothesisRanking[0]?.posteriorProbability ?? 0.5;
  const runnerUpPosterior =
    analysis.hypothesisRanking[1]?.posteriorProbability ?? 0;
  const hypothesisCount = analysis.hypothesisRanking.length;
  const unsupportedClaimRatio =
    hypothesisCount === 0
      ? 0
      : analysis.evidenceAudit.unsupportedClaims.length / hypothesisCount;
  const updateStrength = average(
    analysis.hypothesisRanking.map((entry) =>
      Math.abs(entry.posteriorProbability - entry.priorProbability)
    )
  );

  return {
    hypothesisCount,
    topPosterior,
    runnerUpPosterior,
    evidenceTraceability: analysis.evidenceAudit.traceabilityCoverage,
    unsupportedClaimRatio,
    validationStrength: deriveProbabilityValidationStrength(
      analysis.validation
    ),
    updateStrength,
    convergenceRate: clampUnit(analysis.diagnostics.convergence.observedRate),
    source: '@a0n/maybe probability analysis',
  };
}

/**
 * Handles the gnosis derive Probability Formal Ledger Bundle workflow.
 */
export function deriveProbabilityFormalLedgerBundle(
  analysis: FormalLedgerProbabilityAnalysis,
  declaration?: FormalLedgerServiceDeclaration
): FormalLedgerBundle {
  return deriveFormalLedgerBundle(
    buildFormalLedgerContext(declaration, {
      probability: createProbabilityFormalLedgerInput(analysis),
    })
  );
}

/**
 * Creates the Metacognition Formal Ledger Context.
 */
export function createMetacognitionFormalLedgerContext(
  analysis: FormalLedgerMetacognitionAnalysis,
  declaration?: FormalLedgerServiceDeclaration
): FormalLedgerContext {
  if (!hasMetacognitionAnalysisSignal(analysis)) {
    return buildFormalLedgerContext(declaration);
  }

  const calibrationLayers = analysis.calibration.layers ?? [];
  const rawConfidenceValues = calibrationLayers
    .map((layer) => layer.rawConfidence)
    .filter((value): value is number => Number.isFinite(value));
  const calibratedConfidenceValues = calibrationLayers
    .map((layer) => layer.calibratedConfidence)
    .filter((value): value is number => Number.isFinite(value));
  const dataQualityValues = calibrationLayers
    .map((layer) => layer.dataQuality)
    .filter((value): value is number => Number.isFinite(value));
  const credibleIntervalWidths = calibrationLayers
    .map((layer) => {
      if (layer.credibleInterval == null) {
        return undefined;
      }

      return clampUnit(layer.credibleInterval[1] - layer.credibleInterval[0]);
    })
    .filter((value): value is number => Number.isFinite(value));
  const layerAlignmentScores = (analysis.coherence.layerAlignmentScores ?? [])
    .filter((value): value is number => Number.isFinite(value))
    .map(clampUnit);
  const rawConfidence = averageOrDefault(
    rawConfidenceValues,
    clampUnit(analysis.calibration.overallConfidence ?? 0.5)
  );
  const calibratedConfidence = averageOrDefault(
    calibratedConfidenceValues,
    rawConfidence
  );
  const dataQuality = averageOrDefault(dataQualityValues, calibratedConfidence);
  const intervalWidth = averageOrDefault(credibleIntervalWidths, 0.35);
  const overallCoherence = clampUnit(
    analysis.coherence.overallCoherence ??
      averageOrDefault(layerAlignmentScores, 0.5)
  );
  const coherenceImprovement = clampUnit(
    analysis.coherence.coherenceImprovement ?? overallCoherence
  );
  const remainingTensionCount = Math.max(
    0,
    Math.round(analysis.coherence.remainingTensionCount ?? 0)
  );
  const resolvedContradictionCount = Math.max(
    0,
    Math.round(analysis.coherence.resolvedContradictionCount ?? 0)
  );
  const totalContradictions =
    remainingTensionCount + resolvedContradictionCount;
  const resolvedRatio =
    totalContradictions === 0
      ? 1
      : clampUnit(resolvedContradictionCount / totalContradictions);
  const layerAlignment = averageOrDefault(
    layerAlignmentScores,
    overallCoherence
  );
  const detectedCount = Math.max(
    0,
    Math.round(analysis.bias.detectedCount ?? 0)
  );
  const correctedCount = Math.max(
    0,
    Math.round(analysis.bias.correctedCount ?? detectedCount)
  );
  const correctedRatio =
    detectedCount === 0 ? 1 : clampUnit(correctedCount / detectedCount);
  const correctionConfidence = clampUnit(
    analysis.bias.confidenceInCorrections ?? 0.5
  );
  const affectedLayerBreadth = clampUnit(
    analysis.bias.affectedLayerBreadth ??
      (detectedCount === 0 ? 0.5 : correctedRatio)
  );
  const debiasedPressure = clampUnit(analysis.bias.totalDebiasedDelta ?? 0);
  const overallQuality = clampUnit(
    analysis.quality?.overallQuality ??
      averageOrDefault(
        [calibratedConfidence, overallCoherence, correctionConfidence],
        0.5
      )
  );

  return buildFormalLedgerContext(declaration, {
    metacognition: {
      c0Confidence: averageOrDefault(
        [rawConfidence, dataQuality, overallQuality],
        rawConfidence
      ),
      c1Calibration: averageOrDefault(
        [calibratedConfidence, dataQuality, 1 - intervalWidth, overallQuality],
        calibratedConfidence
      ),
      c2Coherence: averageOrDefault(
        [
          overallCoherence,
          coherenceImprovement,
          layerAlignment,
          resolvedRatio,
          overallQuality,
        ],
        overallCoherence
      ),
      c3BiasAwareness: averageOrDefault(
        [
          correctionConfidence,
          correctedRatio,
          affectedLayerBreadth,
          1 - debiasedPressure,
          overallQuality,
        ],
        correctionConfidence
      ),
      source: analysis.source ?? 'gnosis metacognition analysis',
    },
  });
}

/**
 * Handles the gnosis derive Metacognition Formal Ledger Bundle workflow.
 */
export function deriveMetacognitionFormalLedgerBundle(
  analysis: FormalLedgerMetacognitionAnalysis,
  declaration?: FormalLedgerServiceDeclaration
): FormalLedgerBundle {
  return deriveFormalLedgerBundle(
    createMetacognitionFormalLedgerContext(analysis, declaration)
  );
}

/**
 * Creates the Recovery Formal Ledger Context.
 */
export function createRecoveryFormalLedgerContext(
  analysis: FormalLedgerRecoveryAnalysis,
  declaration?: FormalLedgerServiceDeclaration
): FormalLedgerContext {
  if (!hasRecoveryAnalysisSignal(analysis)) {
    return buildFormalLedgerContext(declaration);
  }

  const indicators = analysis.markers?.indicators ?? [];
  const recoveryFactors = analysis.markers?.recoveryFactors ?? [];
  const confidence = clampUnit(analysis.signals.confidence ?? 0.5);
  const burnoutRisk = clampUnit(analysis.signals.burnoutRisk ?? 0.5);
  const recoveryTimeLoad =
    typeof analysis.signals.recoveryTimeMinutes === 'number' &&
    Number.isFinite(analysis.signals.recoveryTimeMinutes)
      ? normalizeCount(Math.max(0, analysis.signals.recoveryTimeMinutes), 240)
      : averageOrDefault([burnoutRisk, confidence], 0.5);
  const stressLoad = clampUnit(
    analysis.signals.stressLoad ??
      averageOrDefault([burnoutRisk, recoveryTimeLoad], 0.5)
  );
  const indicatorLoad = normalizeCount(indicators.length, 4);
  const recoveryFactorSupport = normalizeCount(recoveryFactors.length, 4);
  const breathSignal = keywordSignalScore(recoveryFactors, [
    'breath',
    'breathing',
    'exhale',
    'inhale',
    'regulat',
    'calm',
  ]);
  const restSignal = keywordSignalScore(recoveryFactors, [
    'sleep',
    'rest',
    'recover',
    'restore',
    'break',
    'pause',
  ]);
  const forgivenessSignal = keywordSignalScore(recoveryFactors, [
    'forgive',
    'release',
    'accept',
    'grace',
    'compassion',
  ]);
  const ruminationSignal = keywordSignalScore(indicators, [
    'ruminat',
    'replay',
    'spiral',
    'loop',
    'catastroph',
    'obsess',
    'defensive',
  ]);
  const restoration = clampUnit(
    analysis.signals.restorationCapacity ??
      averageOrDefault(
        [
          1 - stressLoad,
          1 - burnoutRisk,
          recoveryFactorSupport,
          restSignal,
          confidence,
        ],
        0.5
      )
  );
  const sleepDebt = clampUnit(
    analysis.signals.sleepDebt ??
      averageOrDefault(
        [recoveryTimeLoad, burnoutRisk, 1 - restSignal, indicatorLoad],
        0.5
      )
  );
  const forgiveness = clampUnit(
    analysis.signals.forgiveness ??
      averageOrDefault(
        [forgivenessSignal, restoration, 1 - ruminationSignal],
        0.5
      )
  );
  const breathRegulation = clampUnit(
    analysis.signals.breathRegulation ??
      averageOrDefault([breathSignal, 1 - stressLoad, confidence], 0.45)
  );
  const rumination = clampUnit(
    analysis.signals.ruminationRisk ??
      averageOrDefault(
        [ruminationSignal, stressLoad, recoveryTimeLoad, indicatorLoad],
        0.5
      )
  );

  return buildFormalLedgerContext(declaration, {
    recovery: {
      stressLoad,
      sleepDebt,
      restoration,
      forgiveness,
      breathRegulation,
      rumination,
      source: analysis.source ?? 'gnosis recovery analysis',
    },
  });
}

/**
 * Handles the gnosis derive Recovery Formal Ledger Bundle workflow.
 */
export function deriveRecoveryFormalLedgerBundle(
  analysis: FormalLedgerRecoveryAnalysis,
  declaration?: FormalLedgerServiceDeclaration
): FormalLedgerBundle {
  return deriveFormalLedgerBundle(
    createRecoveryFormalLedgerContext(analysis, declaration)
  );
}

/**
 * Creates the Justice Formal Ledger Context.
 */
export function createJusticeFormalLedgerContext(
  analysis: FormalLedgerJusticeAnalysis,
  declaration?: FormalLedgerServiceDeclaration
): FormalLedgerContext {
  if (!hasJusticeAnalysisSignal(analysis)) {
    return buildFormalLedgerContext(declaration);
  }

  const evidence = analysis.evidence;
  const adjudication = analysis.adjudication;
  const process = analysis.process;
  const totalThreads = Math.max(
    1,
    Math.round(
      evidence.totalThreads ??
        Math.max(
          0,
          (evidence.coveredThreads ?? 0) +
            (evidence.contestedThreads ?? 0) +
            (evidence.uncoveredThreads ?? 0)
        )
    )
  );
  const coveredThreads = Math.max(
    0,
    evidence.coveredThreads ?? totalThreads - (evidence.uncoveredThreads ?? 0)
  );
  const contestedThreads = Math.max(0, evidence.contestedThreads ?? 0);
  const uncoveredThreads = Math.max(
    0,
    evidence.uncoveredThreads ??
      Math.max(totalThreads - coveredThreads - contestedThreads, 0)
  );
  const coverageRatio = clampUnit(coveredThreads / totalThreads);
  const contestedRatio = clampUnit(contestedThreads / totalThreads);
  const uncoveredRatio = clampUnit(uncoveredThreads / totalThreads);
  const evidenceDeficit = clampUnit(
    evidence.evidenceDeficit == null
      ? averageOrDefault([uncoveredRatio, contestedRatio], 0.35)
      : normalizeCount(evidence.evidenceDeficit, Math.max(2, totalThreads))
  );
  const averageConfidence = clampUnit(
    evidence.averageConfidence ??
      averageOrDefault(
        [coverageRatio, 1 - contestedRatio, 1 - uncoveredRatio],
        0.5
      )
  );
  const criticalIssueLoad = normalizeCount(
    process?.criticalIssueCount ?? 0,
    Math.max(2, totalThreads)
  );
  const injusticeTriggerLoad = normalizeCount(
    process?.injusticeTriggerCount ?? 0,
    Math.max(2, totalThreads)
  );
  const remedyCandidateCount = Math.max(
    0,
    adjudication?.remedyCandidateCount ?? adjudication?.rankedRemedyCount ?? 0
  );
  const remedyCountScore = normalizeCount(remedyCandidateCount, 4);
  const recommendedActionCoverage = normalizeCount(
    process?.recommendedActionCount ?? remedyCandidateCount,
    Math.max(2, totalThreads)
  );
  const summaryClarity = clampUnit(
    process?.summaryClarity ??
      averageOrDefault([recommendedActionCoverage, 1 - criticalIssueLoad], 0.5)
  );
  const ratificationRequired = adjudication?.ratificationRequired === true;
  const enforcementActive = adjudication?.enforcementActive === true;
  const appealAvailable = adjudication?.appealAvailable === true;
  const communitySignalBreadth = normalizeCount(
    adjudication?.communitySignalCount ??
      adjudication?.affectedCommunityCount ??
      0,
    4
  );
  const readiness = (() => {
    if (evidence.sufficient === true) {
      return 0.9;
    }

    switch (adjudication?.status) {
      case 'adjudicated_remedy':
      case 'mediated_settlement':
        return 0.82;
      case 'ratification_required':
        return 0.58;
      default:
        return 0.28;
    }
  })();
  const selectedRemedyConfidence =
    adjudication?.status === 'adjudicated_remedy' ||
    adjudication?.status === 'mediated_settlement'
      ? 0.82
      : adjudication?.status === 'ratification_required'
      ? 0.58
      : 0.35;
  const dominanceRisk = averageOrDefault(
    [
      criticalIssueLoad,
      injusticeTriggerLoad,
      uncoveredRatio,
      enforcementActive ? 0.78 : 0.34,
    ],
    0.5
  );

  return buildFormalLedgerContext(declaration, {
    probability: {
      hypothesisCount: Math.max(2, remedyCandidateCount || totalThreads),
      topPosterior: averageOrDefault(
        [coverageRatio, averageConfidence, readiness, 1 - criticalIssueLoad],
        0.5
      ),
      runnerUpPosterior: averageOrDefault(
        [
          contestedRatio,
          uncoveredRatio,
          injusticeTriggerLoad,
          ratificationRequired ? 0.64 : 0.28,
        ],
        0.35
      ),
      evidenceTraceability: averageOrDefault(
        [coverageRatio, averageConfidence, 1 - uncoveredRatio, summaryClarity],
        0.5
      ),
      unsupportedClaimRatio: averageOrDefault(
        [uncoveredRatio, criticalIssueLoad],
        0.25
      ),
      validationStrength: averageOrDefault(
        [
          1 - evidenceDeficit,
          readiness,
          summaryClarity,
          recommendedActionCoverage,
        ],
        0.5
      ),
      updateStrength: averageOrDefault(
        [
          1 - criticalIssueLoad,
          1 - injusticeTriggerLoad,
          recommendedActionCoverage,
          appealAvailable ? 0.72 : 0.46,
        ],
        0.5
      ),
      convergenceRate: averageOrDefault(
        [readiness, 1 - contestedRatio, remedyCountScore],
        0.5
      ),
      source: analysis.source ?? 'gnosis justice evidence surface',
    },
    community: {
      empathy: averageOrDefault(
        [
          communitySignalBreadth,
          recommendedActionCoverage,
          1 - criticalIssueLoad,
        ],
        0.5
      ),
      belonging: averageOrDefault(
        [
          communitySignalBreadth,
          1 - injusticeTriggerLoad,
          enforcementActive ? 0.32 : 0.68,
        ],
        0.5
      ),
      dominanceRisk,
      mergePressure: averageOrDefault(
        [
          ratificationRequired ? 0.78 : 0.42,
          criticalIssueLoad,
          1 - summaryClarity,
        ],
        0.5
      ),
      reciprocity: averageOrDefault(
        [
          recommendedActionCoverage,
          coverageRatio,
          communitySignalBreadth,
          appealAvailable ? 0.68 : 0.45,
        ],
        0.5
      ),
      source: analysis.source ?? 'gnosis justice community field',
    },
    metacognition: {
      c0Confidence: averageOrDefault([averageConfidence, coverageRatio], 0.5),
      c1Calibration: averageOrDefault(
        [coverageRatio, 1 - evidenceDeficit, summaryClarity],
        0.5
      ),
      c2Coherence: averageOrDefault(
        [
          1 - contestedRatio,
          1 - criticalIssueLoad,
          recommendedActionCoverage,
          selectedRemedyConfidence,
        ],
        0.5
      ),
      c3BiasAwareness: averageOrDefault(
        [
          appealAvailable ? 0.78 : 0.42,
          ratificationRequired ? 0.72 : 0.48,
          communitySignalBreadth,
        ],
        0.5
      ),
      source: analysis.source ?? 'gnosis justice metacognitive audit',
    },
  });
}

/**
 * Handles the gnosis derive Justice Formal Ledger Bundle workflow.
 */
export function deriveJusticeFormalLedgerBundle(
  analysis: FormalLedgerJusticeAnalysis,
  declaration?: FormalLedgerServiceDeclaration
): FormalLedgerBundle {
  return deriveFormalLedgerBundle(
    createJusticeFormalLedgerContext(analysis, declaration)
  );
}

/**
 * Creates the Governance Formal Ledger Context.
 */
export function createGovernanceFormalLedgerContext(
  analysis: FormalLedgerGovernanceAnalysis,
  declaration?: FormalLedgerServiceDeclaration
): FormalLedgerContext {
  if (!hasGovernanceAnalysisSignal(analysis)) {
    return buildFormalLedgerContext(declaration);
  }

  const totalCommunities = Math.max(
    1,
    Math.round(deriveGovernanceTotalCommunities(analysis))
  );
  const winner = analysis.winner ?? analysis.rankings[0] ?? null;

  if (winner == null) {
    return buildFormalLedgerContext(declaration);
  }

  const runnerUp =
    analysis.rankings.find(
      (ranking) => ranking.candidateId !== winner.candidateId
    ) ?? analysis.rankings[1];
  const consensusLevel = clampUnit(
    analysis.process?.consensusLevel ??
      averageOrDefault(
        [
          normalizeGovernanceApproval(winner.approvals, totalCommunities),
          1 - normalizeGovernanceDeficit(winner, totalCommunities),
        ],
        0.5
      )
  );
  const dissentRatio = normalizeCount(
    analysis.process?.dissentCount ?? 0,
    totalCommunities
  );
  const emotionalCost = governanceEmotionalCostScore(
    analysis.process?.emotionalCostIndex
  );
  const delegationAvailable = analysis.process?.delegationAvailable === true;
  const condorcetWinner = analysis.process?.condorcetWinner === true;
  const winnerApprovals = normalizeGovernanceApproval(
    winner.approvals,
    totalCommunities
  );
  const winnerRejections = normalizeGovernanceRejections(
    winner.rejections,
    totalCommunities
  );
  const winnerRepresented = normalizeGovernanceCoverage(
    winner.representedCommunities ?? winner.approvals,
    totalCommunities
  );
  const winnerUncovered = normalizeGovernanceCoverage(
    winner.uncoveredCommunities,
    totalCommunities
  );
  const winnerDeficit = normalizeGovernanceDeficit(winner, totalCommunities);
  const leadershipSeparation = deriveGovernanceLeadershipSeparation(
    winner,
    runnerUp,
    totalCommunities,
    condorcetWinner
  );
  const dominanceRisk = averageOrDefault(
    [winnerRejections, winnerDeficit, 1 - winnerRepresented],
    0.5
  );
  const candidateDiversity =
    analysis.rankings.length <= 1
      ? 0.45
      : normalizeCount(analysis.rankings.length - 1, 4);
  const recoveryInput =
    emotionalCost > 0.4 || dissentRatio > 0
      ? {
          stressLoad: averageOrDefault(
            [emotionalCost, dissentRatio, 1 - consensusLevel],
            0.5
          ),
          sleepDebt: averageOrDefault([emotionalCost, dissentRatio], 0.45),
          restoration: averageOrDefault(
            [1 - emotionalCost, consensusLevel],
            0.5
          ),
          forgiveness: averageOrDefault(
            [1 - winnerRejections, delegationAvailable ? 0.64 : 0.52],
            0.5
          ),
          breathRegulation: averageOrDefault(
            [1 - emotionalCost, 1 - dissentRatio],
            0.5
          ),
          rumination: averageOrDefault(
            [dissentRatio, emotionalCost, 1 - leadershipSeparation],
            0.5
          ),
          source: analysis.source ?? 'gnosis governance recovery field',
        }
      : undefined;

  return buildFormalLedgerContext(declaration, {
    negotiation: {
      assertiveness: averageOrDefault(
        [winnerApprovals, leadershipSeparation, condorcetWinner ? 0.76 : 0.56],
        0.5
      ),
      cooperation: averageOrDefault(
        [consensusLevel, winnerRepresented, 1 - emotionalCost],
        0.5
      ),
      batnaStrength: averageOrDefault(
        [
          leadershipSeparation,
          winnerRepresented,
          delegationAvailable ? 0.7 : 0.5,
        ],
        0.5
      ),
      watnaRisk: averageOrDefault(
        [winnerDeficit, winnerRejections, emotionalCost],
        0.5
      ),
      settlementReadiness: averageOrDefault(
        [consensusLevel, 1 - winnerDeficit, delegationAvailable ? 0.72 : 0.56],
        0.5
      ),
      trustReserve: averageOrDefault(
        [winnerRepresented, 1 - dissentRatio, 1 - emotionalCost],
        0.5
      ),
      source: analysis.source ?? 'gnosis governance negotiation field',
    },
    community: {
      empathy: averageOrDefault(
        [winnerRepresented, consensusLevel, delegationAvailable ? 0.68 : 0.52],
        0.5
      ),
      belonging: averageOrDefault(
        [winnerRepresented, 1 - winnerUncovered, 1 - emotionalCost],
        0.5
      ),
      dominanceRisk,
      mergePressure: averageOrDefault(
        [1 - consensusLevel, emotionalCost, dissentRatio],
        0.5
      ),
      reciprocity: averageOrDefault(
        [winnerApprovals, 1 - winnerRejections, leadershipSeparation],
        0.5
      ),
      source: analysis.source ?? 'gnosis governance community field',
    },
    metacognition: {
      c0Confidence: averageOrDefault(
        [consensusLevel, leadershipSeparation, winnerApprovals],
        0.5
      ),
      c1Calibration: averageOrDefault(
        [winnerRepresented, 1 - winnerDeficit, leadershipSeparation],
        0.5
      ),
      c2Coherence: averageOrDefault(
        [consensusLevel, 1 - dissentRatio, 1 - emotionalCost],
        0.5
      ),
      c3BiasAwareness: averageOrDefault(
        [
          delegationAvailable ? 0.78 : 0.46,
          candidateDiversity,
          winnerRepresented,
        ],
        0.5
      ),
      source: analysis.source ?? 'gnosis governance deliberation state',
    },
    ...(recoveryInput == null ? {} : { recovery: recoveryInput }),
  });
}

/**
 * Handles the gnosis derive Governance Formal Ledger Bundle workflow.
 */
export function deriveGovernanceFormalLedgerBundle(
  analysis: FormalLedgerGovernanceAnalysis,
  declaration?: FormalLedgerServiceDeclaration
): FormalLedgerBundle {
  return deriveFormalLedgerBundle(
    createGovernanceFormalLedgerContext(analysis, declaration)
  );
}

/**
 * Creates the Negotiation Formal Ledger Context.
 */
export function createNegotiationFormalLedgerContext(
  analysis: FormalLedgerNegotiationAnalysis,
  declaration?: FormalLedgerServiceDeclaration
): FormalLedgerContext {
  const primary =
    NEGOTIATION_MODE_SCORES[analysis.tkiLikeSummary.primaryMode] ??
    NEGOTIATION_MODE_SCORES.compromising;
  const secondary =
    NEGOTIATION_MODE_SCORES[analysis.tkiLikeSummary.secondaryMode] ??
    NEGOTIATION_MODE_SCORES.compromising;
  const alignmentValues = [
    ...analysis.bazaar.rounds.map((round) => round.alignment),
    ...(analysis.bazaar.trajectory?.map((step) => step.alignment) ?? []),
    analysis.bazaar.strategicPressure?.alignment,
  ].filter((value): value is number => Number.isFinite(value));
  const alignmentAverage = averageOrDefault(alignmentValues, 0.55);
  const escalationRisk = clampUnit(
    analysis.bazaar.strategicPressure?.escalationRisk ?? 0.5
  );
  const settlementPressure = clampUnit(
    analysis.bazaar.strategicPressure?.settlementPressure ?? 0.5
  );
  const volatility = clampUnit(
    analysis.bazaar.strategicPressure?.volatility ?? 0.5
  );
  const neutralConfidence = clampUnit(
    analysis.neutral.settlementCandidates[0]?.confidence ?? 0.55
  );
  const pressureRisk =
    NEGOTIATION_PRESSURE_BIAS_RISK[
      analysis.neutralSettlementBias.pressureBias ?? 'hold'
    ];

  return buildFormalLedgerContext(declaration, {
    negotiation: {
      assertiveness: averageOrDefault(
        [primary.assertiveness, secondary.assertiveness],
        0.5
      ),
      cooperation: averageOrDefault(
        [primary.cooperation, secondary.cooperation],
        0.5
      ),
      batnaStrength: averageOrDefault(
        [
          (analysis.bazaar.branchOutcomes?.length ?? 0) > 0 ? 0.65 : 0.4,
          alignmentAverage,
          analysis.bazaar.settled ? 0.7 : 0.45,
        ],
        0.5
      ),
      watnaRisk: averageOrDefault(
        [pressureRisk, escalationRisk, 1 - alignmentAverage],
        0.5
      ),
      settlementReadiness: averageOrDefault(
        [
          neutralConfidence,
          settlementPressure,
          analysis.neutral.convergenceStatus === 'settled' ? 0.9 : 0.35,
        ],
        0.5
      ),
      trustReserve: averageOrDefault([alignmentAverage, 1 - volatility], 0.5),
      source: 'aeon bazaar + neutral profile',
    },
    community: {
      empathy: averageOrDefault([alignmentAverage, neutralConfidence], 0.5),
      belonging:
        analysis.neutral.convergenceStatus === 'settled'
          ? averageOrDefault([neutralConfidence, 0.8], 0.7)
          : averageOrDefault([neutralConfidence, 0.45], 0.45),
      dominanceRisk: averageOrDefault([pressureRisk, escalationRisk], 0.5),
      mergePressure: averageOrDefault(
        [settlementPressure, neutralConfidence],
        0.5
      ),
      reciprocity: averageOrDefault([alignmentAverage, 1 - volatility], 0.5),
      source: 'aeon negotiation convergence field',
    },
  });
}

/**
 * Handles the gnosis derive Negotiation Formal Ledger Bundle workflow.
 */
export function deriveNegotiationFormalLedgerBundle(
  analysis: FormalLedgerNegotiationAnalysis,
  declaration?: FormalLedgerServiceDeclaration
): FormalLedgerBundle {
  return deriveFormalLedgerBundle(
    createNegotiationFormalLedgerContext(analysis, declaration)
  );
}

/**
 * Creates the Communication Formal Ledger Context.
 */
export function createCommunicationFormalLedgerContext(
  analysis: FormalLedgerCommunicationAnalysis,
  declaration?: FormalLedgerServiceDeclaration
): FormalLedgerContext {
  const rawOutputStreams = analysis.topology.outputStreams;
  const participantCount = Number.isFinite(analysis.topology.participantCount)
    ? Math.max(1, Math.round(analysis.topology.participantCount))
    : 1;
  const outputStreams = (() => {
    if (
      typeof rawOutputStreams === 'number' &&
      Number.isFinite(rawOutputStreams)
    ) {
      return Math.max(1, Math.round(rawOutputStreams));
    }

    return 1;
  })();
  const rawDeficit = Math.max(0, participantCount - outputStreams);
  const agreementRatio = clampUnit(analysis.topology.agreementRatio ?? 0.5);
  const sharedContextUnits = Math.max(
    0,
    analysis.topology.sharedContextUnits ?? rawDeficit * agreementRatio
  );
  const effectiveDeficit = Math.max(0, rawDeficit - sharedContextUnits);
  const unresolvedDeficit =
    rawDeficit === 0 ? 0 : clampUnit(effectiveDeficit / rawDeficit);
  const contextCoverage =
    rawDeficit === 0 ? 1 : clampUnit(1 - unresolvedDeficit);
  const outputCompression =
    participantCount <= 1
      ? 0
      : clampUnit((participantCount - outputStreams) / participantCount);
  const alerts = analysis.safety?.alerts ?? [];
  const alertLoad = averageOrDefault(
    alerts.map((alert) => communicationAlertSeverityScore(alert.severity)),
    analysis.safety?.entropyLevel ?? 0.5
  );
  const actedAlertRatio =
    alerts.length === 0
      ? 0.5
      : clampUnit(
          alerts.filter((alert) => alert.twinTookAction === true).length /
            alerts.length
        );
  const entropyLevel = clampUnit(analysis.safety?.entropyLevel ?? alertLoad);
  const empathySignal = clampUnit(
    analysis.signals.empathySignal ??
      averageOrDefault([agreementRatio, contextCoverage], 0.5)
  );
  const authenticity = clampUnit(
    analysis.signals.authenticity ??
      averageOrDefault([agreementRatio, 1 - entropyLevel, contextCoverage], 0.5)
  );
  const clarificationNeed = clampUnit(
    analysis.signals.clarificationNeed ??
      averageOrDefault(
        [unresolvedDeficit, 1 - agreementRatio, entropyLevel],
        0.5
      )
  );
  const dominanceRisk = clampUnit(
    analysis.signals.dominanceRisk ??
      averageOrDefault([unresolvedDeficit, alertLoad, entropyLevel], 0.5)
  );
  const mergePressure = clampUnit(
    analysis.signals.mergePressure ??
      averageOrDefault(
        [outputCompression, entropyLevel, unresolvedDeficit],
        0.5
      )
  );
  const reciprocitySignal = clampUnit(
    analysis.signals.reciprocitySignal ??
      averageOrDefault(
        [agreementRatio, 1 - dominanceRisk, actedAlertRatio],
        0.5
      )
  );
  const perspectiveBreadth = clampUnit(
    analysis.signals.perspectiveBreadth ??
      averageOrDefault(
        [empathySignal, 1 - dominanceRisk, 1 - clarificationNeed],
        0.5
      )
  );
  const source = analysis.source ?? 'gnosis communication field';

  const community: FormalLedgerCommunityInput = {
    empathy: averageOrDefault([empathySignal, contextCoverage], 0.5),
    belonging: averageOrDefault(
      [authenticity, contextCoverage, 1 - mergePressure],
      0.5
    ),
    dominanceRisk: averageOrDefault([dominanceRisk, alertLoad], 0.5),
    mergePressure: averageOrDefault([mergePressure, unresolvedDeficit], 0.5),
    reciprocity: averageOrDefault(
      [reciprocitySignal, actedAlertRatio, 1 - clarificationNeed],
      0.5
    ),
    source,
  };

  const metacognition: FormalLedgerMetacognitionInput = {
    c0Confidence: averageOrDefault([authenticity, agreementRatio], 0.5),
    c1Calibration: averageOrDefault(
      [1 - clarificationNeed, authenticity, 1 - alertLoad],
      0.5
    ),
    c2Coherence: averageOrDefault(
      [contextCoverage, authenticity, 1 - dominanceRisk],
      0.5
    ),
    c3BiasAwareness: averageOrDefault(
      [perspectiveBreadth, empathySignal, 1 - dominanceRisk],
      0.5
    ),
    source,
  };

  const recoveryInput = createCommunicationRecoveryInput(
    analysis,
    source,
    entropyLevel,
    alertLoad,
    unresolvedDeficit,
    clarificationNeed,
    dominanceRisk,
    authenticity,
    contextCoverage
  );

  return buildFormalLedgerContext(declaration, {
    community,
    metacognition,
    ...(recoveryInput == null ? {} : { recovery: recoveryInput }),
  });
}

/**
 * Handles the gnosis derive Communication Formal Ledger Bundle workflow.
 */
export function deriveCommunicationFormalLedgerBundle(
  analysis: FormalLedgerCommunicationAnalysis,
  declaration?: FormalLedgerServiceDeclaration
): FormalLedgerBundle {
  return deriveFormalLedgerBundle(
    createCommunicationFormalLedgerContext(analysis, declaration)
  );
}

/**
 * Creates the Trust Formal Ledger Context.
 */
export function createTrustFormalLedgerContext(
  analysis: FormalLedgerTrustAnalysis,
  declaration?: FormalLedgerServiceDeclaration
): FormalLedgerContext {
  if (!hasTrustAnalysisSignal(analysis)) {
    return buildFormalLedgerContext(declaration);
  }

  const verifiedActor =
    analysis.authority?.verifiedActor == null
      ? 0.5
      : analysis.authority.verifiedActor
      ? 0.88
      : 0.18;
  const delegationDepth = normalizeCount(
    Math.max(0, analysis.authority?.delegationDepth ?? 0),
    4
  );
  const capabilityCoverage = clampUnit(
    analysis.authority?.capabilityCoverage ??
      averageOrDefault([verifiedActor, 1 - delegationDepth], 0.5)
  );
  const overridePressure = clampUnit(
    analysis.authority?.overridePressure ??
      averageOrDefault([1 - capabilityCoverage, delegationDepth], 0.45)
  );
  const holdCount = normalizeCount(
    Math.max(0, analysis.authority?.holdCount ?? 0),
    4
  );
  const activeHoldRatio = clampUnit(
    analysis.authority?.activeHoldRatio ??
      averageOrDefault([holdCount, overridePressure], holdCount)
  );
  const quarantineRisk = clampUnit(
    analysis.authority?.quarantineRisk ??
      averageOrDefault([activeHoldRatio, overridePressure], 0.4)
  );
  const forgeHoldRisk = clampUnit(
    analysis.authority?.forgeHoldRisk ??
      averageOrDefault([activeHoldRatio, holdCount], 0.35)
  );
  const manipulationLoad = clampUnit(
    analysis.signals.manipulationLoad ??
      averageOrDefault([overridePressure, activeHoldRatio], 0.5)
  );
  const factualCoverage = clampUnit(
    analysis.signals.factualCoverage ??
      averageOrDefault(
        [1 - manipulationLoad, capabilityCoverage, verifiedActor],
        0.5
      )
  );
  const requestClarity = clampUnit(
    analysis.signals.requestClarity ??
      averageOrDefault([factualCoverage, 1 - manipulationLoad], 0.5)
  );
  const impliedPressure = clampUnit(
    analysis.signals.impliedPressure ??
      averageOrDefault([manipulationLoad, 1 - requestClarity], 0.5)
  );
  const authenticityGap = clampUnit(
    analysis.signals.authenticityGap ??
      averageOrDefault([manipulationLoad, 1 - factualCoverage], 0.5)
  );
  const evidenceBreadth = clampUnit(
    analysis.signals.evidenceBreadth ??
      averageOrDefault([factualCoverage, requestClarity], 0.5)
  );
  const authorityStrain = averageOrDefault(
    [overridePressure, activeHoldRatio, quarantineRisk, forgeHoldRisk],
    0.5
  );
  const source = analysis.source ?? 'gnosis trust analysis';

  return buildFormalLedgerContext(declaration, {
    community: {
      empathy: averageOrDefault(
        [factualCoverage, requestClarity, 1 - manipulationLoad],
        0.5
      ),
      belonging: averageOrDefault(
        [verifiedActor, capabilityCoverage, 1 - authorityStrain],
        0.5
      ),
      dominanceRisk: averageOrDefault(
        [overridePressure, manipulationLoad, impliedPressure],
        0.5
      ),
      mergePressure: averageOrDefault(
        [impliedPressure, delegationDepth, activeHoldRatio],
        0.5
      ),
      reciprocity: averageOrDefault(
        [capabilityCoverage, factualCoverage, 1 - manipulationLoad],
        0.5
      ),
      source,
    },
    metacognition: {
      c0Confidence: averageOrDefault(
        [verifiedActor, factualCoverage, capabilityCoverage],
        0.5
      ),
      c1Calibration: averageOrDefault(
        [requestClarity, evidenceBreadth, 1 - authenticityGap],
        0.5
      ),
      c2Coherence: averageOrDefault(
        [factualCoverage, 1 - activeHoldRatio, capabilityCoverage],
        0.5
      ),
      c3BiasAwareness: averageOrDefault(
        [evidenceBreadth, 1 - manipulationLoad, 1 - impliedPressure],
        0.5
      ),
      source,
    },
    recovery: {
      stressLoad: averageOrDefault(
        [manipulationLoad, authorityStrain, impliedPressure],
        0.5
      ),
      sleepDebt: averageOrDefault(
        [overridePressure, activeHoldRatio, quarantineRisk],
        0.45
      ),
      restoration: averageOrDefault(
        [1 - authorityStrain, capabilityCoverage, verifiedActor],
        0.5
      ),
      forgiveness: averageOrDefault(
        [1 - authenticityGap, factualCoverage, 1 - overridePressure],
        0.5
      ),
      breathRegulation: averageOrDefault(
        [1 - manipulationLoad, 1 - impliedPressure, 1 - authorityStrain],
        0.45
      ),
      rumination: averageOrDefault(
        [manipulationLoad, activeHoldRatio, authenticityGap],
        0.5
      ),
      source,
    },
  });
}

/**
 * Handles the gnosis derive Trust Formal Ledger Bundle workflow.
 */
export function deriveTrustFormalLedgerBundle(
  analysis: FormalLedgerTrustAnalysis,
  declaration?: FormalLedgerServiceDeclaration
): FormalLedgerBundle {
  return deriveFormalLedgerBundle(
    createTrustFormalLedgerContext(analysis, declaration)
  );
}

/**
 * Creates the Behavior Formal Ledger Context.
 */
export function createBehaviorFormalLedgerContext(
  analysis: FormalLedgerBehaviorAnalysis,
  declaration?: FormalLedgerServiceDeclaration
): FormalLedgerContext {
  if (!hasBehaviorAnalysisSignal(analysis)) {
    return buildFormalLedgerContext(declaration);
  }

  const confidences = analysis.detections
    .map((detection) => detection.confidence)
    .filter((value): value is number => Number.isFinite(value))
    .map(clampUnit);
  const negativeConfidences = analysis.detections
    .filter((detection) => detection.valence === 'negative')
    .map((detection) => detection.confidence)
    .filter((value): value is number => Number.isFinite(value))
    .map(clampUnit);
  const positiveConfidences = analysis.detections
    .filter((detection) => detection.valence === 'positive')
    .map((detection) => detection.confidence)
    .filter((value): value is number => Number.isFinite(value))
    .map(clampUnit);
  const neutralConfidences = analysis.detections
    .filter((detection) => detection.valence === 'neutral')
    .map((detection) => detection.confidence)
    .filter((value): value is number => Number.isFinite(value))
    .map(clampUnit);
  const matchedTermLoads = analysis.detections
    .map((detection) => detection.matchedTermCount)
    .filter((value): value is number => Number.isFinite(value))
    .map((value) => normalizeCount(Math.max(0, value), 5));
  const detectionDensity = normalizeCount(analysis.detections.length, 5);
  const interventionCoverage =
    analysis.detections.length === 0
      ? normalizeCount(analysis.summary?.actionableCount ?? 0, 3)
      : clampUnit(
          analysis.detections.filter(
            (detection) => detection.hasIntervention === true
          ).length / analysis.detections.length
        );
  const recommendationCoverage = normalizeCount(
    analysis.recommendationCount ?? analysis.summary?.actionableCount ?? 0,
    4
  );
  const topConfidence = averageOrDefault(confidences, 0.5);
  const negativeConfidence = averageOrDefault(negativeConfidences, 0);
  const positiveConfidence = averageOrDefault(positiveConfidences, 0);
  const neutralConfidence = averageOrDefault(neutralConfidences, 0);
  const matchedTermLoad = averageOrDefault(matchedTermLoads, detectionDensity);
  const riskLoad = behaviorRiskLevelScore(analysis.summary?.riskLevel);
  const dominantValencePressure = behaviorDominantValenceScore(
    analysis.summary?.dominantValence
  );
  const loopInstability = averageOrDefault(
    [detectionDensity, matchedTermLoad, riskLoad],
    0.5
  );
  const negativeLoad = averageOrDefault(
    [negativeConfidence, riskLoad, dominantValencePressure.negative],
    0.5
  );
  const positiveSupport = averageOrDefault(
    [
      positiveConfidence,
      recommendationCoverage,
      interventionCoverage,
      dominantValencePressure.positive,
    ],
    0.4
  );
  const ambiguityLoad = averageOrDefault(
    [
      neutralConfidence,
      dominantValencePressure.ambiguous,
      1 - topConfidence,
      1 - recommendationCoverage,
    ],
    0.45
  );
  const dominanceRisk = averageOrDefault(
    [negativeLoad, riskLoad, loopInstability],
    0.5
  );
  const source = analysis.source ?? 'gnosis behavior analysis';

  return buildFormalLedgerContext(declaration, {
    community: {
      empathy: averageOrDefault(
        [positiveSupport, 1 - negativeLoad, interventionCoverage],
        0.5
      ),
      belonging: averageOrDefault(
        [positiveSupport, 1 - negativeLoad, 1 - ambiguityLoad],
        0.5
      ),
      dominanceRisk,
      mergePressure: averageOrDefault(
        [loopInstability, ambiguityLoad, negativeLoad],
        0.5
      ),
      reciprocity: averageOrDefault(
        [interventionCoverage, recommendationCoverage, 1 - negativeLoad],
        0.5
      ),
      source,
    },
    metacognition: {
      c0Confidence: averageOrDefault(
        [topConfidence, recommendationCoverage, 1 - ambiguityLoad],
        0.5
      ),
      c1Calibration: averageOrDefault(
        [interventionCoverage, matchedTermLoad, 1 - ambiguityLoad],
        0.5
      ),
      c2Coherence: averageOrDefault(
        [1 - loopInstability, positiveSupport, 1 - negativeLoad],
        0.5
      ),
      c3BiasAwareness: averageOrDefault(
        [recommendationCoverage, interventionCoverage, 1 - negativeLoad],
        0.5
      ),
      source,
    },
    recovery: {
      stressLoad: averageOrDefault(
        [negativeLoad, riskLoad, loopInstability],
        0.5
      ),
      sleepDebt: averageOrDefault(
        [negativeLoad, ambiguityLoad, detectionDensity],
        0.45
      ),
      restoration: averageOrDefault(
        [positiveSupport, interventionCoverage, 1 - negativeLoad],
        0.5
      ),
      forgiveness: averageOrDefault(
        [positiveSupport, 1 - dominanceRisk, 1 - negativeLoad],
        0.5
      ),
      breathRegulation: averageOrDefault(
        [1 - negativeLoad, interventionCoverage, 1 - riskLoad],
        0.45
      ),
      rumination: averageOrDefault(
        [negativeLoad, ambiguityLoad, detectionDensity],
        0.5
      ),
      source,
    },
  });
}

/**
 * Handles the gnosis derive Behavior Formal Ledger Bundle workflow.
 */
export function deriveBehaviorFormalLedgerBundle(
  analysis: FormalLedgerBehaviorAnalysis,
  declaration?: FormalLedgerServiceDeclaration
): FormalLedgerBundle {
  return deriveFormalLedgerBundle(
    createBehaviorFormalLedgerContext(analysis, declaration)
  );
}

function createVoidWalkingReport(
  input?: FormalLedgerPersonalityInput
): FormalLedgerFamilyReport {
  if (input == null || !hasAnyDefinedValue(input)) {
    return createInactiveReport('voidWalking');
  }

  const primitives = {
    fork: clampUnit(input.try_ ?? INV_PHI),
    race: clampUnit(input.choose ?? INV_PHI),
    fold: clampUnit(input.commit ?? INV_PHI),
    vent: clampUnit(input.letGo ?? INV_PHI),
    interfere: clampUnit(input.learn ?? INV_PHI),
  } satisfies Record<FormalLedgerPrimitiveId, number>;
  const states = toPrimitiveStates(primitives, {
    fork: 'Alternatives remain live long enough to make a real choice.',
    race: 'Selection criteria keep signal and noise separate.',
    fold: 'Commitment survives friction long enough to close the loop.',
    vent: 'Stress is released without erasing the lesson.',
    interfere: 'Reflection changes the chooser instead of just narrating it.',
  });
  const spike = pickPrimitive(states, 'highest');
  const scaffold = pickPrimitive(states, 'lowest');
  const balanceScore = calculateBalanceScore(states);
  const summary = `${
    spike.title
  } carries the heaviest current deficit while ${scaffold.title.toLowerCase()} remains the best scaffold.`;

  return createActiveReport(
    'voidWalking',
    states,
    summary,
    [
      createDiagnostic(
        'voidWalking',
        'spike',
        'Current spike',
        `${spike.title} is the largest visible deficit in the five-primitive profile.`,
        spike.severity,
        spike.id,
        spike.deficit
      ),
      createDiagnostic(
        'voidWalking',
        'balance',
        'Convergence balance',
        `Balance score is ${balanceScore.toFixed(
          2
        )} around the phi-adjacent attractor reference.`,
        balanceScore < 0.45 ? 'high' : balanceScore < 0.7 ? 'moderate' : 'low',
        undefined,
        balanceScore
      ),
    ],
    [
      {
        id: 'voidWalking-wellness',
        familyId: 'voidWalking',
        audience: 'wellness',
        title: 'Void-walking profile',
        summary: `${
          spike.title
        } needs the most care right now. Stabilize that axis first, then use ${scaffold.title.toLowerCase()} as the recovery scaffold.`,
        detail:
          spike.id === 'vent'
            ? 'Regulation before interpretation. Exhale-heavy recovery, sleep protection, and gentle letting-go practices reduce vent debt fastest.'
            : `Use ${spike.title.toLowerCase()} as the primary intervention axis and keep ${scaffold.title.toLowerCase()} visible as the most stable foothold.`,
      },
      {
        id: 'voidWalking-research',
        familyId: 'voidWalking',
        audience: 'research',
        title: 'Five-Bule mapping',
        summary:
          'Try/Choose/Commit/LetGo/Learn remains the executable product surface for fork/race/fold/vent/interfere.',
        detail:
          'Source lineage: THM-FIVE-BULE-PERSONALITY, THM-FIVE-MAP-TO-FIVE, THM-CONVERGENCE-UNIFICATION, and the executable @a0n/buleyean-rl personality model.',
      },
    ],
    createVoidWalkingRecommendations(spike),
    [
      createUnmappedSegment(
        'source',
        'Personality source',
        1,
        input.source ?? 'Seven-layer or quark-derived profile'
      ),
    ],
    {
      attractor: INV_PHI,
      balanceScore,
      spike: spike.id,
      scaffold: scaffold.id,
    }
  );
}

function createNegotiationReport(
  input?: FormalLedgerNegotiationInput
): FormalLedgerFamilyReport {
  if (input == null || !hasAnyDefinedValue(input)) {
    return createInactiveReport('negotiation');
  }

  const assertiveness = clampUnit(input.assertiveness ?? 0.5);
  const cooperation = clampUnit(input.cooperation ?? 0.5);
  const batnaStrength = clampUnit(input.batnaStrength ?? 0.5);
  const watnaRisk = clampUnit(input.watnaRisk ?? 0.5);
  const settlementReadiness = clampUnit(input.settlementReadiness ?? 0.5);
  const trustReserve = clampUnit(input.trustReserve ?? 0.5);
  const balance = 1 - Math.abs(assertiveness - cooperation);

  const primitives = {
    fork: average([batnaStrength, 1 - watnaRisk]),
    race: average([settlementReadiness, cooperation]),
    fold: average([settlementReadiness, balance]),
    vent: average([1 - watnaRisk, trustReserve]),
    interfere: average([cooperation, trustReserve, balance]),
  } satisfies Record<FormalLedgerPrimitiveId, number>;
  const states = toPrimitiveStates(primitives, {
    fork: 'Options stay live long enough to avoid false urgency.',
    race: 'Selection pressure is bounded by criteria instead of panic.',
    fold: 'Closure can happen without pretending the field is cleaner than it is.',
    vent: 'Failed branches can be released without escalating repair debt.',
    interfere:
      'The negotiation updates itself instead of looping the same threat pattern.',
  });
  const spike = pickPrimitive(states, 'highest');
  const summary = `${
    spike.title
  } is the leading negotiation deficit. BATNA/WATNA tension is ${watnaRisk.toFixed(
    2
  )} vs ${batnaStrength.toFixed(2)}.`;

  const diagnostics: FormalLedgerDiagnostic[] = [
    createDiagnostic(
      'negotiation',
      'batna-watna',
      'BATNA/WATNA spread',
      `BATNA strength ${batnaStrength.toFixed(
        2
      )} and WATNA risk ${watnaRisk.toFixed(
        2
      )} set the current bargaining ceiling.`,
      watnaRisk > batnaStrength ? 'high' : 'moderate',
      'fork',
      batnaStrength - watnaRisk
    ),
  ];

  if (balance < 0.55) {
    diagnostics.push(
      createDiagnostic(
        'negotiation',
        'mode-imbalance',
        'Mode imbalance',
        'Assertiveness and cooperation are too far apart to settle cleanly.',
        'moderate',
        'fold',
        balance
      )
    );
  }

  return createActiveReport(
    'negotiation',
    states,
    summary,
    diagnostics,
    [
      {
        id: 'negotiation-wellness',
        familyId: 'negotiation',
        audience: 'wellness',
        title: 'Negotiation posture',
        summary:
          watnaRisk > 0.6
            ? 'The field is reading as unsafe. Bound the negotiation before asking it to converge.'
            : 'The field is still workable. Tighten criteria and move toward a bounded settlement surface.',
        detail:
          'Bazaar keeps multiplicity alive; Neutral is the bounded landing zone when the open field starts charging repair debt.',
      },
      {
        id: 'negotiation-research',
        familyId: 'negotiation',
        audience: 'research',
        title: 'Bazaar / Neutral routing',
        summary:
          'NegotiationEquilibrium, aeon-bazaar, and aeon-neutral form the bounded-unbounded routing surface for this family.',
        detail:
          'High WATNA pressure is the signal to switch from open bazaar exploration toward neutral mediation and convergence certificates.',
      },
    ],
    createNegotiationRecommendations(
      watnaRisk,
      batnaStrength,
      settlementReadiness
    ),
    [
      createUnmappedSegment(
        'assertiveness',
        'Assertiveness',
        assertiveness,
        'Negotiation style input that shapes closure pressure but is not itself a primitive.'
      ),
      createUnmappedSegment(
        'cooperation',
        'Cooperation',
        cooperation,
        'Counterparty alignment input that shapes selection and update quality.'
      ),
    ],
    {
      watnaRisk,
      batnaStrength,
      settlementReadiness,
      trustReserve,
    }
  );
}

function createCommunityReport(
  input?: FormalLedgerCommunityInput
): FormalLedgerFamilyReport {
  if (input == null || !hasAnyDefinedValue(input)) {
    return createInactiveReport('community');
  }

  const empathy = clampUnit(input.empathy ?? 0.5);
  const belonging = clampUnit(input.belonging ?? 0.5);
  const dominanceRisk = clampUnit(input.dominanceRisk ?? 0.5);
  const mergePressure = clampUnit(input.mergePressure ?? 0.5);
  const reciprocity = clampUnit(input.reciprocity ?? 0.5);

  const primitives = {
    fork: empathy,
    race: reciprocity,
    fold: average([belonging, reciprocity]),
    vent: average([1 - dominanceRisk, 1 - mergePressure]),
    interfere: average([empathy, reciprocity, 1 - dominanceRisk]),
  } satisfies Record<FormalLedgerPrimitiveId, number>;
  const states = toPrimitiveStates(primitives, {
    fork: 'Other paths remain legible instead of being collapsed into one imposed story.',
    race: 'Communal signal can be selected without flattening dissent.',
    fold: 'Belonging survives convergence instead of being bought by domination.',
    vent: 'Pressure can discharge without becoming social debt or coercion.',
    interfere:
      'Communities update from one another instead of entrenching blind spots.',
  });
  const summary = `Dominance risk ${dominanceRisk.toFixed(
    2
  )} and merge pressure ${mergePressure.toFixed(
    2
  )} define the current community surface.`;

  return createActiveReport(
    'community',
    states,
    summary,
    [
      createDiagnostic(
        'community',
        'dominance',
        'Dominance pressure',
        'Merged-community health is threatened when one lane starts closing everybody else’s option field.',
        dominanceRisk > 0.6 ? 'high' : 'moderate',
        'vent',
        dominanceRisk
      ),
      createDiagnostic(
        'community',
        'merge',
        'Merge pressure',
        'Merge pressure is only safe when reciprocity and belonging stay above coercion.',
        mergePressure > 0.6 ? 'high' : 'moderate',
        'fold',
        mergePressure
      ),
    ],
    [
      {
        id: 'community-wellness',
        familyId: 'community',
        audience: 'wellness',
        title: 'Community field',
        summary:
          dominanceRisk > 0.6
            ? 'Restore turn-taking and veto space before asking the group to merge harder.'
            : 'The group can converge if reciprocity stays visible and pressure remains dischargeable.',
        detail:
          'Empathy convergence is a group health signal, not a demand for sameness.',
      },
      {
        id: 'community-research',
        familyId: 'community',
        audience: 'research',
        title: 'Community dominance surface',
        summary:
          'CommunityDominance and SkyrmsNadirBule provide the bounded language for attenuation, dominance, and merge deficits.',
        detail:
          'This family treats merged-community failure as a convergence problem: diversity can amplify or collapse depending on whether dominance remains bounded.',
      },
    ],
    createCommunityRecommendations(dominanceRisk, mergePressure, reciprocity),
    [
      createUnmappedSegment(
        'belonging',
        'Belonging',
        belonging,
        'A social stability measure that shapes fold quality but is not itself a primitive.'
      ),
    ],
    {
      empathy,
      belonging,
      dominanceRisk,
      mergePressure,
      reciprocity,
    }
  );
}

function createMetacognitionReport(
  input?: FormalLedgerMetacognitionInput
): FormalLedgerFamilyReport {
  if (input == null || !hasAnyDefinedValue(input)) {
    return createInactiveReport('metacognition');
  }

  const c0Confidence = clampUnit(input.c0Confidence ?? 0.5);
  const c1Calibration = clampUnit(input.c1Calibration ?? 0.5);
  const c2Coherence = clampUnit(input.c2Coherence ?? 0.5);
  const c3BiasAwareness = clampUnit(input.c3BiasAwareness ?? 0.5);

  const primitives = {
    fork: average([c0Confidence, c3BiasAwareness]),
    race: c1Calibration,
    fold: c2Coherence,
    vent: c3BiasAwareness,
    interfere: average([c1Calibration, c2Coherence, c3BiasAwareness]),
  } satisfies Record<FormalLedgerPrimitiveId, number>;
  const states = toPrimitiveStates(primitives, {
    fork: 'Observation remains broad enough to preserve alternate explanations.',
    race: 'Calibration separates confidence from wishful certainty.',
    fold: 'Contradictions are reconciled instead of being stacked indefinitely.',
    vent: 'Old self-stories can be released when better evidence arrives.',
    interfere:
      'The system learns from its own mistakes instead of repeating them with new language.',
  });

  return createActiveReport(
    'metacognition',
    states,
    `C0-C3 is currently anchored by calibration ${c1Calibration.toFixed(
      2
    )} and coherence ${c2Coherence.toFixed(2)}.`,
    [
      createDiagnostic(
        'metacognition',
        'calibration',
        'Calibration gap',
        'Confidence should not outrun calibration.',
        c1Calibration < 0.45 ? 'high' : 'moderate',
        'race',
        c1Calibration
      ),
      createDiagnostic(
        'metacognition',
        'coherence',
        'Coherence check',
        'Cross-layer contradictions should be surfaced before they are narrated as certainty.',
        c2Coherence < 0.45 ? 'high' : 'moderate',
        'fold',
        c2Coherence
      ),
    ],
    [
      {
        id: 'metacognition-wellness',
        familyId: 'metacognition',
        audience: 'wellness',
        title: 'Metacognitive state',
        summary:
          c2Coherence < 0.45
            ? 'Slow down and reconcile contradictions before treating the current self-story as settled.'
            : 'Your reflection loop is usable, but it still benefits from explicit evidence and bias checks.',
        detail:
          'Observation is not the same thing as certainty. C1-C3 exists to keep those apart.',
      },
      {
        id: 'metacognition-research',
        familyId: 'metacognition',
        audience: 'research',
        title: 'Daisy-chain state',
        summary:
          'MetacognitiveDaisyChain supplies the C0-C3 convergence language for calibration, coherence, and bias-aware update.',
        detail:
          'This family is the platform’s theorem-backed bridge from raw profile prediction to self-correcting personality reconstruction.',
      },
    ],
    createMetacognitionRecommendations(
      c1Calibration,
      c2Coherence,
      c3BiasAwareness
    ),
    [
      createUnmappedSegment(
        'c0',
        'C0 confidence',
        c0Confidence,
        'Base-model confidence shapes observation coverage before higher-order checks.'
      ),
    ],
    {
      c0Confidence,
      c1Calibration,
      c2Coherence,
      c3BiasAwareness,
    }
  );
}

function createRecoveryReport(
  input?: FormalLedgerRecoveryInput
): FormalLedgerFamilyReport {
  if (input == null || !hasAnyDefinedValue(input)) {
    return createInactiveReport('recovery');
  }

  const stressLoad = clampUnit(input.stressLoad ?? 0.5);
  const sleepDebt = clampUnit(input.sleepDebt ?? 0.5);
  const restoration = clampUnit(input.restoration ?? 0.5);
  const forgiveness = clampUnit(input.forgiveness ?? 0.5);
  const breathRegulation = clampUnit(input.breathRegulation ?? 0.5);
  const rumination = clampUnit(input.rumination ?? 0.5);

  const primitives = {
    fork: average([1 - stressLoad, forgiveness]),
    race: average([breathRegulation, 1 - rumination]),
    fold: average([restoration, 1 - stressLoad]),
    vent: average([
      restoration,
      forgiveness,
      breathRegulation,
      1 - sleepDebt,
      1 - rumination,
    ]),
    interfere: average([breathRegulation, restoration, 1 - rumination]),
  } satisfies Record<FormalLedgerPrimitiveId, number>;
  const states = toPrimitiveStates(primitives, {
    fork: 'The system still has room to see recovery options instead of only threat.',
    race: 'Breath and attention can separate urgent noise from the next useful action.',
    fold: 'Recovery habits can stabilize instead of dissolving under load.',
    vent: 'Stress can leave the system without erasing signal or storing extra debt.',
    interfere:
      'The body and story update each other instead of escalating together.',
  });

  const diagnostics: FormalLedgerDiagnostic[] = [
    createDiagnostic(
      'recovery',
      'sleep',
      'Sleep debt',
      'Sleep debt is a first-class recovery constraint, not a side issue.',
      sleepDebt > 0.6 ? 'high' : 'moderate',
      'vent',
      sleepDebt
    ),
    createDiagnostic(
      'recovery',
      'rumination',
      'Rumination load',
      'Rumination keeps vent debt live long after the original signal should have discharged.',
      rumination > 0.6 ? 'high' : 'moderate',
      'race',
      rumination
    ),
  ];

  if (stressLoad > 0.65) {
    diagnostics.push(
      createDiagnostic(
        'recovery',
        'stress',
        'Stress load',
        'Stress is high enough that recovery needs to precede interpretation and productivity.',
        'high',
        'fold',
        stressLoad
      )
    );
  }

  return createActiveReport(
    'recovery',
    states,
    `Vent debt is being set by sleep debt ${sleepDebt.toFixed(
      2
    )}, stress ${stressLoad.toFixed(2)}, and rumination ${rumination.toFixed(
      2
    )}.`,
    diagnostics,
    [
      {
        id: 'recovery-wellness',
        familyId: 'recovery',
        audience: 'wellness',
        title: 'Recovery posture',
        summary:
          sleepDebt > 0.6 || stressLoad > 0.65
            ? 'Protect sleep and regulate first. Insight work is downstream of basic recovery.'
            : 'Recovery is available, but it still depends on discharging stress instead of rehearsing it.',
        detail:
          'Breath, sleep, and forgiving release are the fast path for lowering vent debt in this family.',
      },
      {
        id: 'recovery-research',
        familyId: 'recovery',
        audience: 'research',
        title: 'Sleep-debt / vent surface',
        summary:
          'SleepDebt and vent-restoration mappings are treated as theorem-backed recovery constraints, not wellness garnish.',
        detail:
          'This family is the formal bridge from Chapter 17 mindfulness language into service-ready recovery diagnostics.',
      },
    ],
    createRecoveryRecommendations(sleepDebt, stressLoad, rumination),
    [
      createUnmappedSegment(
        'stressLoad',
        'Stress load',
        stressLoad,
        'Stress is a driver of vent debt and fold instability, not a primitive itself.'
      ),
    ],
    {
      stressLoad,
      sleepDebt,
      restoration,
      forgiveness,
      breathRegulation,
      rumination,
    }
  );
}

function createProbabilityReport(
  input?: FormalLedgerProbabilityInput
): FormalLedgerFamilyReport {
  if (input == null || !hasAnyDefinedValue(input)) {
    return createInactiveReport('probability');
  }

  const hypothesisBreadth = normalizeCount(input.hypothesisCount ?? 0, 4);
  const topPosterior = clampUnit(input.topPosterior ?? 0.5);
  const runnerUpPosterior = clampUnit(input.runnerUpPosterior ?? 0.5);
  const posteriorLead = clampUnit(topPosterior - runnerUpPosterior);
  const evidenceTraceability = clampUnit(input.evidenceTraceability ?? 0.5);
  const unsupportedClaimRatio = clampUnit(input.unsupportedClaimRatio ?? 0);
  const validationStrength = clampUnit(input.validationStrength ?? 0.25);
  const updateStrength = clampUnit(input.updateStrength ?? 0);
  const convergenceRate = clampUnit(input.convergenceRate ?? 0);
  const auditStrength = average([
    evidenceTraceability,
    1 - unsupportedClaimRatio,
  ]);

  const primitives = {
    fork: average([hypothesisBreadth, auditStrength]),
    race: average([posteriorLead, validationStrength]),
    fold: average([topPosterior, convergenceRate, validationStrength]),
    vent: average([auditStrength, updateStrength]),
    interfere: average([
      updateStrength,
      validationStrength,
      evidenceTraceability,
    ]),
  } satisfies Record<FormalLedgerPrimitiveId, number>;
  const states = toPrimitiveStates(primitives, {
    fork: 'Multiple live hypotheses remain visible without collapsing the field too early.',
    race: 'The posterior can separate signal from noise instead of leaving every claim equally plausible.',
    fold: 'Inference can close on a working belief state without pretending the evidence is final.',
    vent: 'Unsupported or stale claims can be released instead of being carried as probability debt.',
    interfere:
      'Evidence changes the belief state rather than being absorbed without updating the model.',
  });

  const diagnostics: FormalLedgerDiagnostic[] = [
    createDiagnostic(
      'probability',
      'separation',
      'Posterior separation',
      'The posterior should meaningfully separate the leading hypothesis from the nearest alternative.',
      posteriorLead < 0.12 ? 'high' : posteriorLead < 0.22 ? 'moderate' : 'low',
      'race',
      posteriorLead
    ),
    createDiagnostic(
      'probability',
      'traceability',
      'Evidence traceability',
      'Probability claims should preserve provenance, citations, and calibration lineage.',
      evidenceTraceability < 0.55
        ? 'high'
        : evidenceTraceability < 0.75
        ? 'moderate'
        : 'low',
      'vent',
      evidenceTraceability
    ),
    createDiagnostic(
      'probability',
      'validation',
      'Validation pressure',
      'Posterior confidence should not outrun holdout and negative-control validation.',
      validationStrength < 0.45
        ? 'high'
        : validationStrength < 0.65
        ? 'moderate'
        : 'low',
      'fold',
      validationStrength
    ),
  ];

  if (unsupportedClaimRatio > 0.25) {
    diagnostics.push(
      createDiagnostic(
        'probability',
        'unsupported',
        'Unsupported-claim load',
        'Too much of the probability surface is still being asserted without evidence support.',
        unsupportedClaimRatio > 0.5 ? 'high' : 'moderate',
        'vent',
        unsupportedClaimRatio
      )
    );
  }

  return createActiveReport(
    'probability',
    states,
    `Probability posture is set by posterior lead ${posteriorLead.toFixed(
      2
    )}, traceability ${evidenceTraceability.toFixed(
      2
    )}, and validation ${validationStrength.toFixed(2)}.`,
    diagnostics,
    [
      {
        id: 'probability-wellness',
        familyId: 'probability',
        audience: 'wellness',
        title: 'Probability posture',
        summary:
          posteriorLead < 0.12 || updateStrength < 0.12
            ? 'Keep multiple hypotheses live. The current evidence is not yet separating the field enough to close on one story.'
            : validationStrength < 0.5
            ? 'The posterior is useful for reflection, but not yet strong enough for high-confidence commitment.'
            : 'The field is differentiating cleanly enough to act, as long as evidence lineage remains visible.',
        detail:
          'In Maybe, Solomonoff, Bayesian, frequentist, and retrocausal layers are one walk across the same primitive surface.',
      },
      {
        id: 'probability-research',
        familyId: 'probability',
        audience: 'research',
        title: 'Maybe four-layer state',
        summary:
          'This family formalizes probability as a layered void walk rather than a detached scalar score.',
        detail:
          'Source lineage: @a0n/maybe, the Solomonoff/Bayesian/Frequentist/Retrocausal layer stack, and the Gnosis formal-ledger bridge.',
      },
    ],
    createProbabilityRecommendations(
      posteriorLead,
      evidenceTraceability,
      unsupportedClaimRatio,
      validationStrength,
      updateStrength
    ),
    [
      createUnmappedSegment(
        'hypothesisCount',
        'Hypothesis count',
        hypothesisBreadth,
        'How many live hypotheses are in play before primitive reduction.'
      ),
      createUnmappedSegment(
        'topPosterior',
        'Top posterior',
        topPosterior,
        'The current leading posterior mass before it is reduced to primitive health.'
      ),
      createUnmappedSegment(
        'runnerUpPosterior',
        'Runner-up posterior',
        runnerUpPosterior,
        'The nearest competing posterior mass in the current field.'
      ),
    ],
    {
      hypothesisCount: input.hypothesisCount ?? 0,
      topPosterior,
      runnerUpPosterior,
      posteriorLead,
      evidenceTraceability,
      unsupportedClaimRatio,
      validationStrength,
      updateStrength,
      convergenceRate,
    }
  );
}

function createCommunicationRecoveryInput(
  analysis: FormalLedgerCommunicationAnalysis,
  source: string,
  entropyLevel: number,
  alertLoad: number,
  unresolvedDeficit: number,
  clarificationNeed: number,
  dominanceRisk: number,
  authenticity: number,
  contextCoverage: number
): FormalLedgerRecoveryInput | undefined {
  const safety = analysis.safety;

  if (
    safety == null ||
    ![
      safety.stressLoad,
      safety.sleepDebt,
      safety.restorationCapacity,
      safety.breathRegulation,
      safety.ruminationRisk,
    ].some((value) => Number.isFinite(value))
  ) {
    return undefined;
  }

  return {
    stressLoad: clampUnit(
      safety.stressLoad ??
        averageOrDefault([entropyLevel, alertLoad, unresolvedDeficit], 0.5)
    ),
    sleepDebt: clampUnit(safety.sleepDebt ?? 0.5),
    restoration: clampUnit(
      safety.restorationCapacity ??
        averageOrDefault([1 - entropyLevel, contextCoverage], 0.5)
    ),
    forgiveness: averageOrDefault(
      [contextCoverage, 1 - dominanceRisk, authenticity],
      0.5
    ),
    breathRegulation: clampUnit(
      safety.breathRegulation ??
        averageOrDefault([1 - entropyLevel, authenticity], 0.5)
    ),
    rumination: clampUnit(
      safety.ruminationRisk ??
        averageOrDefault([clarificationNeed, unresolvedDeficit], 0.5)
    ),
    source,
  };
}

function createVoidWalkingRecommendations(
  spike: FormalLedgerPrimitiveState
): readonly FormalLedgerRecommendation[] {
  const base: Record<FormalLedgerPrimitiveId, FormalLedgerRecommendation> = {
    fork: {
      id: 'voidWalking-fork',
      familyId: 'voidWalking',
      title: 'Widen the live option field',
      rationale:
        'Fork deficits are eased by recovering alternatives before forcing a decision.',
      actions: [
        'List one live option you have been excluding.',
        'Delay irreversible choice until at least one alternate path is named.',
      ],
      modalities: ['reflection', 'mindfulness'],
      priority: 'support',
    },
    race: {
      id: 'voidWalking-race',
      familyId: 'voidWalking',
      title: 'Choose one criterion',
      rationale:
        'Race deficits improve when the chooser stops treating every signal as equally urgent.',
      actions: [
        'Name the single criterion that matters most for this decision.',
        'Discard one criterion that is only adding noise.',
      ],
      modalities: ['reflection'],
      priority: 'stabilize',
    },
    fold: {
      id: 'voidWalking-fold',
      familyId: 'voidWalking',
      title: 'Make the smallest irreversible move',
      rationale:
        'Fold deficits need closure pressure that is real but still bounded.',
      actions: [
        'Take one action that makes the current path more concrete.',
        'Close one escape hatch that is only extending indecision.',
      ],
      modalities: ['behavioral'],
      priority: 'stabilize',
    },
    vent: {
      id: 'voidWalking-vent',
      familyId: 'voidWalking',
      title: 'Regulate before interpretation',
      rationale:
        'Vent deficits respond fastest to exhale-heavy regulation, sleep protection, and letting-go practices.',
      actions: [
        'Protect the next sleep window before adding more analysis.',
        'Use an exhale-heavy breathing cycle before revisiting the story.',
      ],
      modalities: ['breathing', 'sleep', 'mindfulness'],
      priority: 'urgent',
    },
    interfere: {
      id: 'voidWalking-interfere',
      familyId: 'voidWalking',
      title: 'Track body, choice, consequence',
      rationale:
        'Interfere deficits drop when reflection changes future behavior instead of staying abstract.',
      actions: [
        'Write one body cue, one choice, and one consequence from the last loop.',
        'Name one behavior you will update on the next repetition.',
      ],
      modalities: ['reflection', 'mindfulness'],
      priority: 'stabilize',
    },
  };

  return [base[spike.id]];
}

function createNegotiationRecommendations(
  watnaRisk: number,
  batnaStrength: number,
  settlementReadiness: number
): readonly FormalLedgerRecommendation[] {
  const recommendations: FormalLedgerRecommendation[] = [];

  if (watnaRisk > 0.6) {
    recommendations.push({
      id: 'negotiation-bound',
      familyId: 'negotiation',
      title: 'Bound the field',
      rationale:
        'When WATNA dominates, open bargaining should route toward a bounded neutral surface.',
      actions: [
        'State a hard floor before the next round.',
        'Move the next exchange into a bounded mediation structure.',
      ],
      modalities: ['behavioral', 'reflection'],
      priority: 'urgent',
    });
  }

  if (batnaStrength < 0.45) {
    recommendations.push({
      id: 'negotiation-batna',
      familyId: 'negotiation',
      title: 'Strengthen BATNA',
      rationale:
        'Weak alternatives force the negotiation to carry more fear than it can metabolize.',
      actions: [
        'Name one external option that lowers dependency on this outcome.',
        'Avoid making concessions until a fallback path is explicit.',
      ],
      modalities: ['reflection'],
      priority: 'stabilize',
    });
  }

  if (settlementReadiness < 0.55) {
    recommendations.push({
      id: 'negotiation-settlement',
      familyId: 'negotiation',
      title: 'Define the landing zone',
      rationale:
        'Settlement readiness improves when the acceptable outcome is explicit before the next move.',
      actions: [
        'Write the minimum acceptable settlement in one sentence.',
        'Separate preferred outcome from acceptable outcome.',
      ],
      modalities: ['reflection'],
      priority: 'support',
    });
  }

  return recommendations;
}

function createCommunityRecommendations(
  dominanceRisk: number,
  mergePressure: number,
  reciprocity: number
): readonly FormalLedgerRecommendation[] {
  const recommendations: FormalLedgerRecommendation[] = [];

  if (dominanceRisk > 0.6) {
    recommendations.push({
      id: 'community-turn-taking',
      familyId: 'community',
      title: 'Restore veto space',
      rationale:
        'Dominance risk drops when all participants regain room to decline, slow, or reframe.',
      actions: [
        'Add explicit turn-taking and dissent space to the next group decision.',
        'Separate urgency from authority before the next merge request.',
      ],
      modalities: ['behavioral'],
      priority: 'urgent',
    });
  }

  if (mergePressure > 0.55) {
    recommendations.push({
      id: 'community-merge',
      familyId: 'community',
      title: 'Merge more slowly',
      rationale:
        'Communities converge better when reciprocity outruns pressure.',
      actions: [
        'Reduce the number of decisions forced into one merge event.',
        'Pilot convergence on one narrow issue before broad unification.',
      ],
      modalities: ['behavioral'],
      priority: 'stabilize',
    });
  }

  if (reciprocity < 0.5) {
    recommendations.push({
      id: 'community-reciprocity',
      familyId: 'community',
      title: 'Increase mutual visibility',
      rationale:
        'Reciprocity is a precondition for healthy fold, not an optional social extra.',
      actions: [
        'Ask each side to restate the other side’s need before proposing a merge.',
        'Track who has changed their view, not just who has spoken the most.',
      ],
      modalities: ['reflection'],
      priority: 'support',
    });
  }

  return recommendations;
}

function createMetacognitionRecommendations(
  c1Calibration: number,
  c2Coherence: number,
  c3BiasAwareness: number
): readonly FormalLedgerRecommendation[] {
  const recommendations: FormalLedgerRecommendation[] = [];

  if (c1Calibration < 0.5) {
    recommendations.push({
      id: 'metacognition-calibration',
      familyId: 'metacognition',
      title: 'Separate confidence from evidence',
      rationale:
        'Calibration improves when certainty is forced to cite the evidence supporting it.',
      actions: [
        'For each strong claim, write one piece of evidence and one missing piece.',
        'Downgrade any conclusion that cannot name a disconfirming test.',
      ],
      modalities: ['reflection'],
      priority: 'stabilize',
    });
  }

  if (c2Coherence < 0.5) {
    recommendations.push({
      id: 'metacognition-coherence',
      familyId: 'metacognition',
      title: 'Run a contradiction pass',
      rationale:
        'Coherence recovers when conflicting self-claims are made explicit instead of blended together.',
      actions: [
        'Write the two statements that currently cannot both be true.',
        'Name which layer each statement comes from before reconciling them.',
      ],
      modalities: ['reflection'],
      priority: 'urgent',
    });
  }

  if (c3BiasAwareness < 0.5) {
    recommendations.push({
      id: 'metacognition-bias',
      familyId: 'metacognition',
      title: 'Name the likely bias',
      rationale:
        'Bias is easier to reduce once it is named as a candidate failure mode instead of hidden inside the story.',
      actions: [
        'Write one cultural, defensive, or self-serving bias that may be shaping the conclusion.',
        'Ask what the opposite inference would be if that bias were active.',
      ],
      modalities: ['reflection'],
      priority: 'support',
    });
  }

  return recommendations;
}

function createRecoveryRecommendations(
  sleepDebt: number,
  stressLoad: number,
  rumination: number
): readonly FormalLedgerRecommendation[] {
  const recommendations: FormalLedgerRecommendation[] = [];

  if (sleepDebt > 0.6) {
    recommendations.push({
      id: 'recovery-sleep',
      familyId: 'recovery',
      title: 'Protect sleep first',
      rationale:
        'Sleep debt is a hard recovery limiter and compounds vent debt across loops.',
      actions: [
        'Protect the next recovery window before adding more demand.',
        'Reduce pre-sleep stimulation and unfinished decisions tonight.',
      ],
      modalities: ['sleep'],
      priority: 'urgent',
    });
  }

  if (stressLoad > 0.6 || rumination > 0.6) {
    recommendations.push({
      id: 'recovery-breath',
      familyId: 'recovery',
      title: 'Shift physiology before analysis',
      rationale:
        'Stress and rumination drop faster when the body exits alarm before the mind re-enters the story.',
      actions: [
        'Run a slow exhale breathing cycle before revisiting the issue.',
        'Delay interpretation until the body has visibly downshifted.',
      ],
      modalities: ['breathing', 'mindfulness'],
      priority: 'stabilize',
    });
  }

  recommendations.push({
    id: 'recovery-release',
    familyId: 'recovery',
    title: 'Practice bounded release',
    rationale:
      'Vent debt shrinks when the lesson is kept but the extra charge is not.',
    actions: [
      'Name what the loop has already taught you.',
      'Release one additional layer of blame, guilt, or replay that is no longer adding signal.',
    ],
    modalities: ['mindfulness', 'reflection'],
    priority: 'support',
  });

  return recommendations;
}

function createProbabilityRecommendations(
  posteriorLead: number,
  evidenceTraceability: number,
  unsupportedClaimRatio: number,
  validationStrength: number,
  updateStrength: number
): readonly FormalLedgerRecommendation[] {
  const recommendations: FormalLedgerRecommendation[] = [];

  if (evidenceTraceability < 0.65 || unsupportedClaimRatio > 0.25) {
    recommendations.push({
      id: 'probability-audit',
      familyId: 'probability',
      title: 'Audit evidence lineage',
      rationale:
        'Probability claims become hard to trust when provenance, citations, or calibration are missing.',
      actions: [
        'Add provenance or citation coverage to the unsupported claims first.',
        'Mark any uncalibrated evidence as provisional before using it to close the field.',
      ],
      modalities: ['reflection'],
      priority:
        evidenceTraceability < 0.5 || unsupportedClaimRatio > 0.5
          ? 'urgent'
          : 'stabilize',
    });
  }

  if (updateStrength < 0.12 || posteriorLead < 0.12) {
    recommendations.push({
      id: 'probability-differentiate',
      familyId: 'probability',
      title: 'Add differentiating evidence',
      rationale:
        'When priors and posteriors barely move apart, the field still needs evidence that can actually distinguish the live hypotheses.',
      actions: [
        'Collect one observation that would raise one hypothesis while lowering another.',
        'Keep at least two hypotheses open until the next discriminating signal arrives.',
      ],
      modalities: ['reflection', 'behavioral'],
      priority: 'stabilize',
    });
  }

  if (validationStrength < 0.5) {
    recommendations.push({
      id: 'probability-validate',
      familyId: 'probability',
      title: 'Collect labeled outcomes',
      rationale:
        'Holdout and negative-control validation keep posterior confidence from outrunning the evidence base.',
      actions: [
        'Record the next outcome labels or historical observations needed for validation.',
        'Re-run the analysis only after the holdout surface has enough structure to falsify it.',
      ],
      modalities: ['behavioral'],
      priority: 'urgent',
    });
  }

  return recommendations;
}

function createInactiveReport(
  familyId: FormalLedgerFamilyId
): FormalLedgerFamilyReport {
  const catalog = FAMILY_CATALOG[familyId];
  return {
    familyId,
    title: catalog.title,
    active: false,
    summary: 'No input declared for this family.',
    sources: catalog.sources,
    primitives: [],
    diagnostics: [],
    narratives: [],
    recommendations: [],
    visualization: {
      id: `${familyId}-visualization`,
      familyId,
      kind: catalog.kind,
      title: `${catalog.title} visualization`,
      summary: 'No input declared for this family.',
      primitives: [],
      metrics: {
        balanceScore: 0,
        activeDeficit: 0,
        referenceAttractor: INV_PHI,
      },
    },
    unmappedSegments: [],
    metadata: {},
  };
}

function createActiveReport(
  familyId: FormalLedgerFamilyId,
  primitiveStates: readonly FormalLedgerPrimitiveState[],
  summary: string,
  diagnostics: readonly FormalLedgerDiagnostic[],
  narratives: readonly FormalLedgerNarrative[],
  recommendations: readonly FormalLedgerRecommendation[],
  unmappedSegments: readonly FormalLedgerUnmappedSegment[],
  metadata: Readonly<Record<string, string | number | boolean>>
): FormalLedgerFamilyReport {
  const catalog = FAMILY_CATALOG[familyId];
  const spike = pickPrimitive(primitiveStates, 'highest');
  const balanceScore = calculateBalanceScore(primitiveStates);
  return {
    familyId,
    title: catalog.title,
    active: true,
    summary,
    sources: catalog.sources,
    primitives: primitiveStates,
    diagnostics,
    narratives,
    recommendations,
    visualization: {
      id: `${familyId}-visualization`,
      familyId,
      kind: catalog.kind,
      title: `${catalog.title} visualization`,
      summary,
      primitives: primitiveStates,
      highlightedPrimitiveId: spike.id,
      metrics: {
        balanceScore,
        activeDeficit: spike.deficit,
        referenceAttractor: INV_PHI,
      },
    },
    unmappedSegments,
    metadata,
  };
}

function toPrimitiveStates(
  primitives: Readonly<Record<FormalLedgerPrimitiveId, number>>,
  summaries: Readonly<Record<FormalLedgerPrimitiveId, string>>
): readonly FormalLedgerPrimitiveState[] {
  return FORMAL_LEDGER_PRIMITIVES.map((primitiveId) => {
    const strength = clampUnit(primitives[primitiveId]);
    const deficit = clampUnit(1 - strength);
    return {
      id: primitiveId,
      title: PRIMITIVE_TITLES[primitiveId],
      strength,
      deficit,
      severity: classifySeverity(deficit),
      summary: summaries[primitiveId],
    };
  });
}

function aggregateBundlePrimitives(
  reports: Readonly<Record<FormalLedgerFamilyId, FormalLedgerFamilyReport>>,
  activeFamilies: readonly FormalLedgerFamilyId[]
): readonly FormalLedgerPrimitiveState[] {
  if (activeFamilies.length === 0) {
    return FORMAL_LEDGER_PRIMITIVES.map((primitiveId) => ({
      id: primitiveId,
      title: PRIMITIVE_TITLES[primitiveId],
      strength: 0.5,
      deficit: 0.5,
      severity: 'moderate',
      summary: 'No active family has declared this primitive yet.',
    }));
  }

  return FORMAL_LEDGER_PRIMITIVES.map((primitiveId) => {
    const strengths = activeFamilies.map((familyId) => {
      const state = reports[familyId].primitives.find(
        (primitive) => primitive.id === primitiveId
      );
      return state?.strength ?? 0.5;
    });
    const strength = average(strengths);
    const deficit = clampUnit(1 - strength);
    return {
      id: primitiveId,
      title: PRIMITIVE_TITLES[primitiveId],
      strength,
      deficit,
      severity: classifySeverity(deficit),
      summary:
        'Average primitive state across all active formal-ledger families.',
    };
  });
}

function createCompliance(
  declaration: FormalLedgerServiceDeclaration | undefined,
  activeFamilies: readonly FormalLedgerFamilyId[]
): FormalLedgerBundle['compliance'] {
  if (declaration?.optedOut) {
    return {
      declaredFamilies: [],
      undeclaredFamilies: activeFamilies,
      optedOut: true,
      status: 'opted_out',
    };
  }

  const declaredFamilies = Array.from(
    new Set([...(declaration?.consume ?? []), ...(declaration?.emit ?? [])])
  );
  const undeclaredFamilies = activeFamilies.filter(
    (familyId) => !declaredFamilies.includes(familyId)
  );

  return {
    declaredFamilies,
    undeclaredFamilies,
    optedOut: false,
    status: undeclaredFamilies.length === 0 ? 'declared' : 'undeclared',
  };
}

function createComplianceDiagnostic(
  compliance: FormalLedgerBundle['compliance']
): FormalLedgerDiagnostic | undefined {
  if (compliance.status !== 'undeclared') {
    return undefined;
  }

  return {
    id: 'formal-ledger-compliance',
    familyId: 'voidWalking',
    title: 'Formal ledger declaration gap',
    summary: `Active families must be declared explicitly. Missing: ${compliance.undeclaredFamilies.join(
      ', '
    )}.`,
    severity: 'moderate',
    sourceRefs: ['formal-ledger-service-declaration'],
  };
}

function createDiagnostic(
  familyId: FormalLedgerFamilyId,
  suffix: string,
  title: string,
  summary: string,
  severity: FormalLedgerSeverity,
  primitiveId?: FormalLedgerPrimitiveId,
  value?: number
): FormalLedgerDiagnostic {
  return {
    id: `${familyId}-${suffix}`,
    familyId,
    title,
    summary,
    severity,
    sourceRefs: FAMILY_CATALOG[familyId].sources,
    ...(primitiveId !== undefined ? { primitiveId } : {}),
    ...(value !== undefined ? { value } : {}),
  };
}

function createUnmappedSegment(
  id: string,
  label: string,
  value: number,
  summary: string
): FormalLedgerUnmappedSegment {
  return {
    id,
    label,
    value: clampUnit(value),
    summary,
  };
}

function classifySeverity(deficit: number): FormalLedgerSeverity {
  if (deficit >= 0.67) {
    return 'high';
  }
  if (deficit >= 0.34) {
    return 'moderate';
  }
  return 'low';
}

function calculateBalanceScore(
  states: readonly FormalLedgerPrimitiveState[]
): number {
  const deficits = states.map((state) => state.deficit);
  const min = Math.min(...deficits);
  const max = Math.max(...deficits);
  return clampUnit(1 - (max - min));
}

function pickPrimitive(
  states: readonly FormalLedgerPrimitiveState[],
  direction: 'highest' | 'lowest'
): FormalLedgerPrimitiveState {
  const ordered = [...states].sort((left, right) =>
    direction === 'highest'
      ? right.deficit - left.deficit
      : left.deficit - right.deficit
  );
  const selected = ordered[0];
  if (selected == null) {
    throw new Error('Formal ledger requires at least one primitive state');
  }
  return selected;
}

function hasMetacognitionAnalysisSignal(
  analysis: FormalLedgerMetacognitionAnalysis
): boolean {
  if ((analysis.calibration.layers?.length ?? 0) > 0) {
    return true;
  }

  return [
    analysis.calibration.overallConfidence,
    analysis.coherence.overallCoherence,
    analysis.coherence.coherenceImprovement,
    analysis.coherence.remainingTensionCount,
    analysis.coherence.resolvedContradictionCount,
    ...(analysis.coherence.layerAlignmentScores ?? []),
    analysis.bias.detectedCount,
    analysis.bias.correctedCount,
    analysis.bias.confidenceInCorrections,
    analysis.bias.totalDebiasedDelta,
    analysis.bias.affectedLayerBreadth,
    analysis.quality?.overallQuality,
  ].some((value) => typeof value === 'number' && Number.isFinite(value));
}

function hasJusticeAnalysisSignal(
  analysis: FormalLedgerJusticeAnalysis
): boolean {
  return [
    analysis.evidence.totalThreads,
    analysis.evidence.coveredThreads,
    analysis.evidence.contestedThreads,
    analysis.evidence.uncoveredThreads,
    analysis.evidence.evidenceDeficit,
    analysis.evidence.averageConfidence,
    analysis.adjudication?.remedyCandidateCount,
    analysis.adjudication?.rankedRemedyCount,
    analysis.adjudication?.affectedCommunityCount,
    analysis.adjudication?.communitySignalCount,
    analysis.process?.criticalIssueCount,
    analysis.process?.injusticeTriggerCount,
    analysis.process?.recommendedActionCount,
    analysis.process?.summaryClarity,
  ].some((value) => typeof value === 'number' && Number.isFinite(value));
}

function hasGovernanceAnalysisSignal(
  analysis: FormalLedgerGovernanceAnalysis
): boolean {
  if (analysis.rankings.length > 0) {
    return true;
  }

  return [
    analysis.totalCommunities,
    analysis.process?.consensusLevel,
    analysis.process?.dissentCount,
  ].some((value) => typeof value === 'number' && Number.isFinite(value));
}

function hasTrustAnalysisSignal(analysis: FormalLedgerTrustAnalysis): boolean {
  if (typeof analysis.authority?.verifiedActor === 'boolean') {
    return true;
  }

  return [
    analysis.authority?.delegationDepth,
    analysis.authority?.capabilityCoverage,
    analysis.authority?.overridePressure,
    analysis.authority?.holdCount,
    analysis.authority?.activeHoldRatio,
    analysis.authority?.quarantineRisk,
    analysis.authority?.forgeHoldRisk,
    analysis.signals.manipulationLoad,
    analysis.signals.factualCoverage,
    analysis.signals.impliedPressure,
    analysis.signals.authenticityGap,
    analysis.signals.requestClarity,
    analysis.signals.evidenceBreadth,
  ].some((value) => typeof value === 'number' && Number.isFinite(value));
}

function hasBehaviorAnalysisSignal(
  analysis: FormalLedgerBehaviorAnalysis
): boolean {
  if (analysis.detections.length > 0) {
    return true;
  }

  return [analysis.summary?.actionableCount, analysis.recommendationCount].some(
    (value) => typeof value === 'number' && Number.isFinite(value)
  );
}

function hasRecoveryAnalysisSignal(
  analysis: FormalLedgerRecoveryAnalysis
): boolean {
  if ((analysis.markers?.indicators?.length ?? 0) > 0) {
    return true;
  }
  if ((analysis.markers?.recoveryFactors?.length ?? 0) > 0) {
    return true;
  }

  return [
    analysis.signals.stressLoad,
    analysis.signals.sleepDebt,
    analysis.signals.recoveryTimeMinutes,
    analysis.signals.burnoutRisk,
    analysis.signals.restorationCapacity,
    analysis.signals.forgiveness,
    analysis.signals.breathRegulation,
    analysis.signals.ruminationRisk,
    analysis.signals.confidence,
  ].some((value) => typeof value === 'number' && Number.isFinite(value));
}

function average(values: readonly number[]): number {
  if (values.length === 0) {
    return 0;
  }
  const sum = values.reduce((total, value) => total + value, 0);
  return clampUnit(sum / values.length);
}

function averageOrDefault(values: readonly number[], fallback: number): number {
  return values.length === 0 ? fallback : average(values);
}

function communicationAlertSeverityScore(
  severity: FormalLedgerCommunicationAlertSeverity
): number {
  switch (severity) {
    case 'critical':
      return 1;
    case 'high':
      return 0.8;
    case 'medium':
      return 0.55;
    case 'low':
      return 0.3;
  }
}

function keywordSignalScore(
  texts: readonly string[],
  keywords: readonly string[]
): number {
  if (texts.length === 0 || keywords.length === 0) {
    return 0;
  }

  const normalizedTexts = texts
    .map((text) => text.toLowerCase())
    .filter((text) => text.trim() !== '');
  const hits = normalizedTexts.reduce((total, text) => {
    return total + keywords.filter((keyword) => text.includes(keyword)).length;
  }, 0);

  return normalizeCount(hits, Math.max(3, normalizedTexts.length * 2));
}

function normalizeCount(value: number, saturatesAt: number): number {
  if (!Number.isFinite(value) || saturatesAt <= 0) {
    return 0;
  }
  return clampUnit(value / saturatesAt);
}

function deriveGovernanceTotalCommunities(
  analysis: FormalLedgerGovernanceAnalysis
): number {
  if (
    analysis.totalCommunities != null &&
    Number.isFinite(analysis.totalCommunities)
  ) {
    return analysis.totalCommunities;
  }

  const maxima = analysis.rankings.flatMap((ranking) => {
    const represented =
      ranking.representedCommunities ?? ranking.approvals ?? undefined;
    const uncovered = ranking.uncoveredCommunities ?? undefined;
    const totals = [
      represented != null && uncovered != null ? represented + uncovered : 0,
      ranking.approvals != null && ranking.rejections != null
        ? ranking.approvals + ranking.rejections
        : 0,
    ];

    return totals.filter((value) => Number.isFinite(value));
  });

  return Math.max(1, ...maxima, analysis.rankings.length);
}

function normalizeGovernanceCoverage(
  value: number | undefined,
  totalCommunities: number
): number {
  return value == null ? 0.5 : clampUnit(value / Math.max(1, totalCommunities));
}

function normalizeGovernanceApproval(
  approvals: number | undefined,
  totalCommunities: number
): number {
  return approvals == null
    ? 0.5
    : clampUnit(approvals / Math.max(1, totalCommunities));
}

function normalizeGovernanceRejections(
  rejections: number | undefined,
  totalCommunities: number
): number {
  return rejections == null
    ? 0.5
    : clampUnit(rejections / Math.max(1, totalCommunities));
}

function normalizeGovernanceDeficit(
  ranking: FormalLedgerGovernanceAnalysisRanking,
  totalCommunities: number
): number {
  return clampUnit(
    (ranking.deficit ??
      (ranking.rejections ?? 0) + (ranking.uncoveredCommunities ?? 0)) /
      Math.max(1, totalCommunities)
  );
}

function deriveGovernanceLeadershipSeparation(
  winner: FormalLedgerGovernanceAnalysisRanking,
  runnerUp: FormalLedgerGovernanceAnalysisRanking | undefined,
  totalCommunities: number,
  condorcetWinner: boolean
): number {
  if (runnerUp == null) {
    return condorcetWinner ? 0.9 : 0.75;
  }

  const winnerDeficit =
    winner.deficit ??
    (winner.rejections ?? 0) + (winner.uncoveredCommunities ?? 0);
  const runnerUpDeficit =
    runnerUp.deficit ??
    (runnerUp.rejections ?? 0) + (runnerUp.uncoveredCommunities ?? 0);
  const deficitGap = normalizeCount(
    Math.max(runnerUpDeficit - winnerDeficit, 0),
    Math.max(1, Math.ceil(totalCommunities / 2))
  );
  const approvalLead = normalizeCount(
    Math.max((winner.approvals ?? 0) - (runnerUp.approvals ?? 0), 0),
    Math.max(1, totalCommunities)
  );

  return averageOrDefault(
    [deficitGap, approvalLead, condorcetWinner ? 0.82 : 0.55],
    condorcetWinner ? 0.78 : 0.5
  );
}

function governanceEmotionalCostScore(value: string | undefined): number {
  switch (value?.trim().toLowerCase()) {
    case 'high':
      return 0.86;
    case 'medium':
      return 0.58;
    case 'low':
      return 0.24;
    default:
      return 0.45;
  }
}

function behaviorRiskLevelScore(
  value: FormalLedgerBehaviorRiskLevel | undefined
): number {
  switch (value) {
    case 'high':
      return 0.86;
    case 'moderate':
      return 0.58;
    case 'low':
      return 0.24;
    default:
      return 0.45;
  }
}

function behaviorDominantValenceScore(
  value: FormalLedgerBehaviorValence | 'mixed' | undefined
): {
  readonly positive: number;
  readonly negative: number;
  readonly ambiguous: number;
} {
  switch (value) {
    case 'positive':
      return { positive: 0.84, negative: 0.18, ambiguous: 0.24 };
    case 'negative':
      return { positive: 0.18, negative: 0.84, ambiguous: 0.28 };
    case 'mixed':
      return { positive: 0.52, negative: 0.56, ambiguous: 0.76 };
    case 'neutral':
      return { positive: 0.32, negative: 0.32, ambiguous: 0.52 };
    default:
      return { positive: 0.4, negative: 0.4, ambiguous: 0.5 };
  }
}

function deriveProbabilityValidationStrength(
  validation: FormalLedgerProbabilityAnalysisValidation
): number {
  if (validation.skipped) {
    return 0.25;
  }

  const holdoutCorrelation = clampUnit(
    validation.holdout?.correlation.value ?? 0
  );
  const holdoutError = clampUnit(
    1 - (validation.holdout?.meanAbsoluteError.value ?? 1)
  );
  const negativeControl = clampUnit(
    validation.negativeControl?.degradation ?? 0
  );

  return average([holdoutCorrelation, holdoutError, negativeControl]);
}

function clampUnit(value: number): number {
  if (Number.isNaN(value)) {
    return 0;
  }
  if (value < 0) {
    return 0;
  }
  if (value > 1) {
    return 1;
  }
  return Number(value.toFixed(4));
}

function hasAnyDefinedValue(value: Record<string, unknown>): boolean {
  return Object.values(value).some((entry) => entry != null);
}
