// @a0n/gnosis/src/gg-timers.ts
/**
 * gg Timers — default setTimeout/setInterval for .gg topology runtimes.
 *
 * In gg-aeon-shell land, every .gg program runs inside a runtime that
 * installs these as the native timer primitives. A .gg author writes
 * `setTimeout(fn, 5000)` and gets persistent, relay-backed scheduling
 * automatically -- no import, no opt-in.
 *
 * The runtime boot sequence:
 *   1. runtime obtains a DashRelay connection (the PID's own room)
 *   2. calls `installGgTimerBootstrap(relay)` before user topology starts
 *   3. topology code runs; its setTimeout calls are now durable
 *
 * Native timers are still accessible via `nativeSetTimeout` /
 * `nativeSetInterval` for hot-path code that needs microsecond-scale
 * scheduling.
 *
 * This module has NO side effects on import. The bootstrap is explicit
 * so test harnesses and alternate runtimes can skip it.
 *
 * @packageDocumentation
 */

// ─── Native Passthroughs ────────────────────────────────────────────
// Capture natives BEFORE any shim can replace them.
// These are always available under their `native*` names.

const CAPTURED_SET_TIMEOUT = (typeof globalThis !== 'undefined' && typeof globalThis.setTimeout === 'function')
  ? globalThis.setTimeout.bind(globalThis)
  : null;
const CAPTURED_SET_INTERVAL = (typeof globalThis !== 'undefined' && typeof globalThis.setInterval === 'function')
  ? globalThis.setInterval.bind(globalThis)
  : null;
const CAPTURED_CLEAR_TIMEOUT = (typeof globalThis !== 'undefined' && typeof globalThis.clearTimeout === 'function')
  ? globalThis.clearTimeout.bind(globalThis)
  : null;
const CAPTURED_CLEAR_INTERVAL = (typeof globalThis !== 'undefined' && typeof globalThis.clearInterval === 'function')
  ? globalThis.clearInterval.bind(globalThis)
  : null;

/** Always-ephemeral setTimeout (bypasses the durable shim). */
export const nativeSetTimeout: typeof setTimeout = CAPTURED_SET_TIMEOUT as typeof setTimeout;

/** Always-ephemeral setInterval (bypasses the durable shim). */
export const nativeSetInterval: typeof setInterval = CAPTURED_SET_INTERVAL as typeof setInterval;

/** Native clearTimeout (bypasses the durable shim). */
export const nativeClearTimeout: typeof clearTimeout = CAPTURED_CLEAR_TIMEOUT as typeof clearTimeout;

/** Native clearInterval (bypasses the durable shim). */
export const nativeClearInterval: typeof clearInterval = CAPTURED_CLEAR_INTERVAL as typeof clearInterval;

// ─── Relay Contract ─────────────────────────────────────────────────

/**
 * Minimal relay interface the gg timer runtime requires.
 * Any DashRelay-compatible socket satisfies this.
 */
export interface GgTimerRelay {
  send(text: string): Promise<void> | void;
  onMessage(handler: (text: string) => void): () => void;
}

// ─── Registration ───────────────────────────────────────────────────

interface GgTimerRegistration {
  readonly id: string;
  readonly kind: 'timeout' | 'interval';
  readonly periodMs: number;
  handler: (...args: unknown[]) => void;
  args: readonly unknown[];
  cancelled: boolean;
}

const registry = new Map<string, GgTimerRegistration>();

const GG_TIMER_TAG = '__gg-timer-fire__';
const GG_TIMER_PRISM_MIN_MS = 50;
const GG_TIMER_PRISM_MAX_MS = 24 * 60 * 60 * 1000;

function nextId(): string {
  if (typeof crypto !== 'undefined' && typeof crypto.randomUUID === 'function') {
    return `gg-${crypto.randomUUID()}`;
  }
  return `gg-${Date.now()}-${Math.random().toString(36).slice(2)}`;
}

function makeFireEnvelope(id: string, delayMs: number): string {
  const clampedDelay = Math.max(
    GG_TIMER_PRISM_MIN_MS,
    Math.min(GG_TIMER_PRISM_MAX_MS, Math.floor(delayMs))
  );
  return JSON.stringify({
    type: 'prism',
    delayMs: clampedDelay,
    payload: JSON.stringify({ tag: GG_TIMER_TAG, id }),
    payloadKind: 'text',
  });
}

function parseFirePayload(text: string): { id: string } | null {
  if (!text.includes(GG_TIMER_TAG)) return null;
  try {
    const parsed = JSON.parse(text) as { tag?: string; id?: string };
    if (parsed.tag !== GG_TIMER_TAG || typeof parsed.id !== 'string') return null;
    return { id: parsed.id };
  } catch {
    return null;
  }
}

// ─── Bootstrap ──────────────────────────────────────────────────────

interface BootstrapState {
  relay: GgTimerRelay;
  unsubscribe: () => void;
}

let bootstrapState: BootstrapState | null = null;

/**
 * Install the gg durable timer primitives as the global default.
 *
 * Call this at runtime boot before any user topology runs. After install,
 * `globalThis.setTimeout` and `globalThis.setInterval` use the temporal
 * prism for persistence. Calls with delay < 50ms fall through to the
 * native timer (too short to round-trip through the relay).
 *
 * Returns an `uninstall` function that restores the native primitives.
 * Safe to call multiple times -- the second call uninstalls the first.
 */
export function installGgTimerBootstrap(relay: GgTimerRelay): () => void {
  if (bootstrapState) bootstrapState.unsubscribe();

  const unsubMessage = relay.onMessage((text) => {
    const fire = parseFirePayload(text);
    if (!fire) return;
    const reg = registry.get(fire.id);
    if (!reg || reg.cancelled) return;
    try {
      reg.handler(...reg.args);
    } catch {
      // Swallow handler errors -- one bad handler shouldn't break the timer loop
    }
    if (reg.kind === 'timeout') {
      registry.delete(fire.id);
    } else {
      // Interval: schedule the next fire
      void sendFire(relay, reg.id, reg.periodMs);
    }
  });

  // Replace the globals
  const originalSetTimeout = globalThis.setTimeout;
  const originalSetInterval = globalThis.setInterval;
  const originalClearTimeout = globalThis.clearTimeout;
  const originalClearInterval = globalThis.clearInterval;

  globalThis.setTimeout = makeShimSetTimeout(relay);
  globalThis.setInterval = makeShimSetInterval(relay);
  globalThis.clearTimeout = makeShimClear() as typeof clearTimeout;
  globalThis.clearInterval = makeShimClear() as typeof clearInterval;

  const uninstall = (): void => {
    unsubMessage();
    globalThis.setTimeout = originalSetTimeout;
    globalThis.setInterval = originalSetInterval;
    globalThis.clearTimeout = originalClearTimeout;
    globalThis.clearInterval = originalClearInterval;
    registry.clear();
    bootstrapState = null;
  };

  bootstrapState = { relay, unsubscribe: uninstall };
  return uninstall;
}

/** Is the gg timer bootstrap currently active? */
export function isGgTimerBootstrapActive(): boolean {
  return bootstrapState !== null;
}

// ─── Shim Implementation ────────────────────────────────────────────

// Numeric handle mapping: shim returns a number so callers using clearTimeout(id) work.
let handleCounter = 1;
const handleToRegId = new Map<number, string>();

async function sendFire(
  relay: GgTimerRelay,
  id: string,
  delayMs: number
): Promise<void> {
  if (delayMs < GG_TIMER_PRISM_MIN_MS) {
    // Can't durably schedule sub-50ms -- fall back to native
    nativeSetTimeout(() => {
      const reg = registry.get(id);
      if (!reg || reg.cancelled) return;
      try {
        reg.handler(...reg.args);
      } catch {
        // swallow
      }
      if (reg.kind === 'timeout') {
        registry.delete(id);
      } else {
        void sendFire(relay, id, delayMs);
      }
    }, delayMs);
    return;
  }
  const envelope = makeFireEnvelope(id, delayMs);
  await relay.send(envelope);
}

function makeShimSetTimeout(relay: GgTimerRelay): typeof setTimeout {
  return ((
    handler: (...args: unknown[]) => void,
    delay?: number,
    ...args: unknown[]
  ) => {
    const delayMs = typeof delay === 'number' ? delay : 0;
    if (delayMs < GG_TIMER_PRISM_MIN_MS) {
      // Fast path: native
      return nativeSetTimeout(handler as TimerHandler, delayMs, ...args);
    }
    const id = nextId();
    registry.set(id, {
      id,
      kind: 'timeout',
      periodMs: delayMs,
      handler,
      args,
      cancelled: false,
    });
    const numericId = handleCounter++;
    handleToRegId.set(numericId, id);
    void sendFire(relay, id, delayMs);
    return numericId as unknown as ReturnType<typeof setTimeout>;
  }) as typeof setTimeout;
}

function makeShimSetInterval(relay: GgTimerRelay): typeof setInterval {
  return ((
    handler: (...args: unknown[]) => void,
    period?: number,
    ...args: unknown[]
  ) => {
    const periodMs = typeof period === 'number' ? period : 0;
    if (periodMs < GG_TIMER_PRISM_MIN_MS) {
      return nativeSetInterval(handler as TimerHandler, periodMs, ...args);
    }
    const id = nextId();
    registry.set(id, {
      id,
      kind: 'interval',
      periodMs,
      handler,
      args,
      cancelled: false,
    });
    const numericId = handleCounter++;
    handleToRegId.set(numericId, id);
    void sendFire(relay, id, periodMs);
    return numericId as unknown as ReturnType<typeof setInterval>;
  }) as typeof setInterval;
}

function makeShimClear(): (id?: unknown) => void {
  return (id?: unknown): void => {
    if (typeof id !== 'number') {
      // Not one of ours -- try native (they're already captured)
      if (typeof id !== 'undefined' && CAPTURED_CLEAR_TIMEOUT) {
        CAPTURED_CLEAR_TIMEOUT(id as Parameters<typeof clearTimeout>[0]);
      }
      return;
    }
    const regId = handleToRegId.get(id);
    if (regId) {
      const reg = registry.get(regId);
      if (reg) reg.cancelled = true;
      registry.delete(regId);
      handleToRegId.delete(id);
      return;
    }
    // Fall through to native for non-durable handles
    if (CAPTURED_CLEAR_TIMEOUT) {
      CAPTURED_CLEAR_TIMEOUT(id as Parameters<typeof clearTimeout>[0]);
    }
  };
}

/**
 * Diagnostic: current registry size (pending durable timers).
 */
export function ggTimerRegistrySize(): number {
  return registry.size;
}
