// @a0n/gnosis/src/ggc.ts
/**
 * .ggc -- Compiled Gnosis Binary Format
 *
 * PackedKnot binary (0x474B) + CBOR metadata trailer.
 * This is the deployment artifact. Source parsing only in dev.
 *
 * Layout:
 *   [PackedKnot binary...]                  (encode() from packed-knot-encoder)
 *   [4B metadataOffset: uint32 BE]          (offset where metadata starts)
 *   [CBOR metadata: { imports, exports, sourceHash, version }]
 *   [4B magic: 0x47474343 "GGCC"]           (trailer sentinel)
 *
 * The trailer-at-end design means the PackedKnot binary can be
 * extracted by slicing [0..metadataOffset] without parsing CBOR.
 * Fast path: decode knot only. Slow path: also parse metadata.
 *
 * Backed by:
 *   pack_unpack (round-trip identity)
 *   zero_overhead (binary <= source)
 */

import {
  encode as encodeKnot,
  decode as decodeKnot,
  type GgProgram,
} from './packed-knot-encoder';

// ── CBOR (inline minimal encoder/decoder for zero deps) ─────

function encodeCBOR(value: unknown): Uint8Array {
  const parts: number[] = [];
  function writeUint(major: number, n: number): void {
    const mt = major << 5;
    if (n < 24) parts.push(mt | n);
    else if (n < 0x100) parts.push(mt | 24, n);
    else if (n < 0x10000) parts.push(mt | 25, (n >> 8) & 0xff, n & 0xff);
    else
      parts.push(
        mt | 26,
        (n >> 24) & 0xff,
        (n >> 16) & 0xff,
        (n >> 8) & 0xff,
        n & 0xff
      );
  }
  function enc(v: unknown): void {
    if (v === null || v === undefined) {
      parts.push(0xf6);
      return;
    }
    if (v === true) {
      parts.push(0xf5);
      return;
    }
    if (v === false) {
      parts.push(0xf4);
      return;
    }
    if (typeof v === 'number') {
      if (Number.isInteger(v) && v >= 0) writeUint(0, v);
      else if (Number.isInteger(v) && v < 0) writeUint(1, -1 - v);
      else {
        parts.push(0xfb);
        const buf = new ArrayBuffer(8);
        new DataView(buf).setFloat64(0, v);
        parts.push(...new Uint8Array(buf));
      }
      return;
    }
    if (typeof v === 'string') {
      const bytes = new TextEncoder().encode(v);
      writeUint(3, bytes.length);
      parts.push(...bytes);
      return;
    }
    if (v instanceof Uint8Array) {
      writeUint(2, v.length);
      parts.push(...v);
      return;
    }
    if (Array.isArray(v)) {
      writeUint(4, v.length);
      for (const item of v) enc(item);
      return;
    }
    if (typeof v === 'object') {
      const entries = Object.entries(v as Record<string, unknown>);
      writeUint(5, entries.length);
      for (const [k, val] of entries) {
        enc(k);
        enc(val);
      }
      return;
    }
  }
  enc(value);
  return new Uint8Array(parts);
}

function decodeCBOR(bytes: Uint8Array): unknown {
  let offset = 0;
  const readByte = (): number => {
    const byte = bytes[offset];
    if (byte === undefined) {
      throw new Error('Unexpected end of CBOR data');
    }
    offset += 1;
    return byte;
  };
  function read(): unknown {
    const initial = readByte();
    const major = initial >> 5;
    const additional = initial & 0x1f;
    let n: number;
    if (additional < 24) n = additional;
    else if (additional === 24) n = readByte();
    else if (additional === 25) {
      const hi = readByte();
      const lo = readByte();
      n = (hi << 8) | lo;
    } else if (additional === 26) {
      const b0 = readByte();
      const b1 = readByte();
      const b2 = readByte();
      const b3 = readByte();
      n = (b0 << 24) | (b1 << 16) | (b2 << 8) | b3;
    } else n = 0;

    switch (major) {
      case 0:
        return n; // uint
      case 1:
        return -1 - n; // negint
      case 2: {
        const b = bytes.slice(offset, offset + n);
        offset += n;
        return b;
      } // bytes
      case 3: {
        const s = new TextDecoder().decode(bytes.slice(offset, offset + n));
        offset += n;
        return s;
      } // string
      case 4: {
        const arr: unknown[] = [];
        for (let i = 0; i < n; i++) arr.push(read());
        return arr;
      } // array
      case 5: {
        const obj: Record<string, unknown> = {};
        for (let i = 0; i < n; i++) {
          const k = read() as string;
          obj[k] = read();
        }
        return obj;
      } // map
      case 7: {
        if (additional === 20) return false;
        if (additional === 21) return true;
        if (additional === 22) return null;
        if (additional === 27) {
          const end = offset + 8;
          if (end > bytes.length) {
            throw new Error('Unexpected end of CBOR data');
          }
          const buf = bytes.slice(offset, end);
          offset = end;
          return new DataView(buf.buffer, buf.byteOffset).getFloat64(0);
        }
        return null;
      }
    }
    return null;
  }
  return read();
}

// ── GGC Format ──────────────────────────────────────────────

const GGC_TRAILER_MAGIC = 0x47474343; // "GGCC"

export interface GGCMetadata {
  version: number;
  sourceHash: string;
  imports: string[];
  exports: string[];
  compiledAt: string;
  compilerVersion: string;
}

/**
 * Compile a GgProgram + metadata into a .ggc binary.
 */
export function compileToGGC(
  program: GgProgram,
  metadata: Partial<GGCMetadata> = {}
): Uint8Array {
  // Encode the topology as PackedKnot binary
  const knot = encodeKnot(program);

  // Build metadata with defaults
  const meta: GGCMetadata = {
    version: 1,
    sourceHash: metadata.sourceHash ?? '',
    imports: metadata.imports ?? [],
    exports: metadata.exports ?? [],
    compiledAt: metadata.compiledAt ?? new Date().toISOString(),
    compilerVersion: metadata.compilerVersion ?? 'betty/1.0',
  };

  const metaCbor = encodeCBOR(meta);

  // Layout: [knot][CBOR metadata][metadataOffset:4][magic:4]
  // metadataOffset is always at bytes.length - 8
  const metadataOffset = knot.length;
  const total = knot.length + metaCbor.length + 4 + 4;
  const out = new Uint8Array(total);

  // Write knot binary
  out.set(knot, 0);

  // Write CBOR metadata right after knot
  out.set(metaCbor, knot.length);

  // Write metadata offset at total - 8
  const dv = new DataView(out.buffer);
  dv.setUint32(total - 8, metadataOffset, false);

  // Write trailer magic at total - 4
  dv.setUint32(total - 4, GGC_TRAILER_MAGIC, false);

  return out;
}

/**
 * Load a .ggc binary. Returns the decoded program + metadata.
 * Fast path: pass `knotOnly: true` to skip metadata parsing.
 */
export function loadFromGGC(
  bytes: Uint8Array,
  options?: { knotOnly?: boolean }
): { program: GgProgram; metadata: GGCMetadata | null } {
  // Verify trailer magic at last 4 bytes
  const dv = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  const trailerMagic = dv.getUint32(bytes.length - 4, false);
  if (trailerMagic !== GGC_TRAILER_MAGIC) {
    throw new Error('Invalid .ggc file: missing GGCC trailer magic');
  }

  // Layout: [knot:N][CBOR:M][metadataOffset:4][magic:4]
  // metadataOffset at bytes.length - 8
  const mdOffset = dv.getUint32(bytes.length - 8, false);

  // Decode PackedKnot (bytes 0..mdOffset)
  const knotBytes = bytes.slice(0, mdOffset);
  const program = decodeKnot(knotBytes);

  if (options?.knotOnly) {
    return { program, metadata: null };
  }

  // Decode CBOR metadata
  const metaBytes = bytes.slice(mdOffset, bytes.length - 8);
  const metadata = decodeCBOR(metaBytes) as GGCMetadata;

  return { program, metadata };
}

/**
 * Check if a Uint8Array is a valid .ggc file (has the trailer magic).
 */
export function isGGC(bytes: Uint8Array): boolean {
  if (bytes.length < 12) return false;
  const dv = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  return dv.getUint32(bytes.length - 4, false) === GGC_TRAILER_MAGIC;
}

/**
 * Compile a .gg source string to .ggc binary.
 * Convenience function that parses source, compiles, and wraps.
 */
export async function compileSourceToGGC(
  source: string,
  filePath: string,
  options?: { imports?: string[]; exports?: string[] }
): Promise<Uint8Array> {
  // Dynamic import to avoid circular dependency with Betty
  const { BettyCompiler } = await import('./betty/compiler.js');
  const compiler = new BettyCompiler();
  const result = compiler.parse(source);

  if (!result.ast) {
    throw new Error(
      `Compilation failed for ${filePath}: ${result.diagnostics
        ?.map((d) => d.message)
        .join(', ')}`
    );
  }

  // Convert AST to GgProgram format
  const nodes = Array.from(result.ast.nodes.values()).map((n) => ({
    id: n.id,
    labels: n.labels,
    properties: n.properties,
  }));
  const edges = result.ast.edges.map((e) => ({
    sourceIds: e.sourceIds,
    targetIds: e.targetIds,
    type: e.type,
    properties: e.properties,
  }));

  // Hash source for cache validation
  const hashBuffer = await crypto.subtle.digest(
    'SHA-256',
    new TextEncoder().encode(source)
  );
  const sourceHash = Array.from(new Uint8Array(hashBuffer))
    .map((b) => b.toString(16).padStart(2, '0'))
    .join('');

  return compileToGGC(
    { nodes, edges },
    {
      sourceHash,
      imports: options?.imports ?? [],
      exports: options?.exports ?? [],
    }
  );
}
