// @a0n/gnosis/src/gnarly-parser.ts
/**
 * Gnarly parser.
 *
 * `.gnarly` is a GG-family multilingual source format. It keeps GG topology
 * syntax intact, then adds a small metadata block and embedded implementation
 * blocks that the compiler can lower into existing polyglot surfaces.
 */

export interface GnarlyMetadata {
  name?: string;
  languages: string[];
  entry?: string;
  properties: Record<string, string>;
}

export interface GnarlyImplementationBlock {
  nodeId: string;
  language: string;
  sourceText: string;
  startLine: number;
  endLine: number;
}

export interface GnarlyParseDiagnostic {
  line: number;
  column: number;
  message: string;
  severity: 'error' | 'warning' | 'info';
}

export interface GnarlyDocument {
  filePath: string;
  metadata: GnarlyMetadata;
  ggSource: string;
  implementations: GnarlyImplementationBlock[];
  diagnostics: GnarlyParseDiagnostic[];
}

interface SourceRange {
  start: number;
  end: number;
}

const GNARLY_BLOCK_REGEX = /(^|\n)\s*gnarly(?:\s+"([^"]+)")?\s*\{/g;
const IMPL_BLOCK_REGEX =
  /(^|\n)\s*impl\s+([A-Za-z_][A-Za-z0-9_-]*)\s+in\s+([A-Za-z_][A-Za-z0-9_-]*)\s*\{/g;

function lineForIndex(source: string, index: number): number {
  return source.slice(0, index).split('\n').length;
}

function replaceRangeWithBlankLines(source: string, range: SourceRange): string {
  const removed = source.slice(range.start, range.end);
  const replacement = '\n'.repeat(Math.max(1, removed.split('\n').length - 1));
  return `${source.slice(0, range.start)}${replacement}${source.slice(range.end)}`;
}

function findMatchingBrace(source: string, openBraceIndex: number): number {
  let depth = 0;
  for (let index = openBraceIndex; index < source.length; index++) {
    const char = source[index];
    if (char === '{') {
      depth++;
      continue;
    }
    if (char === '}') {
      depth--;
      if (depth === 0) {
        return index;
      }
    }
  }
  return -1;
}

function parseStringList(raw: string): string[] {
  const trimmed = raw.trim();
  if (!trimmed.startsWith('[') || !trimmed.endsWith(']')) {
    return [];
  }
  const body = trimmed.slice(1, -1);
  return body
    .split(',')
    .map((entry) => entry.trim().replace(/^['"]|['"]$/g, ''))
    .filter((entry) => entry.length > 0);
}

function parseMetadataBody(
  body: string
): Pick<GnarlyMetadata, 'languages' | 'entry' | 'properties'> {
  const properties: Record<string, string> = {};
  let languages: string[] = [];
  let entry: string | undefined;

  const assignmentRegex =
    /([A-Za-z_][A-Za-z0-9_-]*)\s*=\s*(\[[^\]]*\]|"[^"]*"|'[^']*'|[^\n,]+)/g;
  let match: RegExpExecArray | null;
  while ((match = assignmentRegex.exec(body)) !== null) {
    const key = match[1] ?? '';
    const rawValue = (match[2] ?? '').trim();
    if (!key) {
      continue;
    }
    if (key === 'languages') {
      languages = parseStringList(rawValue);
      properties[key] = JSON.stringify(languages);
      continue;
    }

    const value = rawValue.replace(/^['"]|['"]$/g, '');
    properties[key] = value;
    if (key === 'entry') {
      entry = value;
    }
  }

  return entry
    ? { languages, entry, properties }
    : { languages, properties };
}

function extractMetadata(
  source: string,
  diagnostics: GnarlyParseDiagnostic[]
): { source: string; metadata: GnarlyMetadata } {
  GNARLY_BLOCK_REGEX.lastIndex = 0;
  const match = GNARLY_BLOCK_REGEX.exec(source);
  if (!match) {
    return {
      source,
      metadata: { languages: [], properties: {} },
    };
  }

  const blockStart = match.index + (match[1]?.length ?? 0);
  const openBraceIndex = source.indexOf('{', blockStart);
  const closeBraceIndex = findMatchingBrace(source, openBraceIndex);
  if (openBraceIndex < 0 || closeBraceIndex < 0) {
    diagnostics.push({
      line: lineForIndex(source, blockStart),
      column: 1,
      message: 'Unclosed gnarly metadata block.',
      severity: 'error',
    });
    const metadata: GnarlyMetadata = { languages: [], properties: {} };
    if (match[2] !== undefined) {
      metadata.name = match[2];
    }
    return { source, metadata };
  }

  const body = source.slice(openBraceIndex + 1, closeBraceIndex);
  const parsed = parseMetadataBody(body);
  const metadata =
    match[2] !== undefined
      ? { name: match[2], ...parsed }
      : { ...parsed };

  return {
    source: replaceRangeWithBlankLines(source, {
      start: blockStart,
      end: closeBraceIndex + 1,
    }),
    metadata,
  };
}

function extractImplementations(
  source: string,
  diagnostics: GnarlyParseDiagnostic[]
): { source: string; implementations: GnarlyImplementationBlock[] } {
  let nextSource = source;
  const implementations: GnarlyImplementationBlock[] = [];
  const ranges: SourceRange[] = [];

  IMPL_BLOCK_REGEX.lastIndex = 0;
  let match: RegExpExecArray | null;
  while ((match = IMPL_BLOCK_REGEX.exec(source)) !== null) {
    const blockStart = match.index + (match[1]?.length ?? 0);
    const openBraceIndex = source.indexOf('{', blockStart);
    const closeBraceIndex = findMatchingBrace(source, openBraceIndex);
    const nodeId = match[2] ?? '';
    const language = match[3] ?? '';

    if (openBraceIndex < 0 || closeBraceIndex < 0) {
      diagnostics.push({
        line: lineForIndex(source, blockStart),
        column: 1,
        message: `Unclosed implementation block for ${nodeId || 'unknown node'}.`,
        severity: 'error',
      });
      continue;
    }

    implementations.push({
      nodeId,
      language,
      sourceText: source.slice(openBraceIndex + 1, closeBraceIndex).trim(),
      startLine: lineForIndex(source, blockStart),
      endLine: lineForIndex(source, closeBraceIndex),
    });
    ranges.push({ start: blockStart, end: closeBraceIndex + 1 });
  }

  for (const range of ranges.sort((a, b) => b.start - a.start)) {
    nextSource = replaceRangeWithBlankLines(nextSource, range);
  }

  return { source: nextSource, implementations };
}

/**
 * Betty parses GG line-by-line. Gnarly allows friendlier multiline node
 * declarations, so compact balanced parenthesized declarations before handing
 * the topology to Betty.
 */
export function normalizeGnarlyGgSource(source: string): string {
  const output: string[] = [];
  let buffer: string[] = [];
  let parenDepth = 0;

  for (const rawLine of source.split('\n')) {
    const line = rawLine.trim();
    if (!line) {
      if (buffer.length === 0) {
        output.push('');
      }
      continue;
    }

    buffer.push(line);
    for (const char of line) {
      if (char === '(') parenDepth++;
      if (char === ')') parenDepth = Math.max(0, parenDepth - 1);
    }

    if (parenDepth === 0) {
      output.push(buffer.join(' '));
      buffer = [];
    }
  }

  if (buffer.length > 0) {
    output.push(buffer.join(' '));
  }

  return output.join('\n').trim();
}

/**
 * Parses the Gnarly.
 */
export function parseGnarly(
  source: string,
  filePath = 'inline.gnarly'
): GnarlyDocument {
  const diagnostics: GnarlyParseDiagnostic[] = [];
  const metadataResult = extractMetadata(source, diagnostics);
  const implementationResult = extractImplementations(
    metadataResult.source,
    diagnostics
  );

  return {
    filePath,
    metadata: metadataResult.metadata,
    ggSource: normalizeGnarlyGgSource(implementationResult.source),
    implementations: implementationResult.implementations,
    diagnostics,
  };
}
