// @a0n/gnosis/src/gnode-execution.ts
import { createHash } from 'node:crypto';
import { BettyCompiler } from './betty/compiler';
import { GnosisCoreCache } from './runtime/core-cache';
import { GnosisEngine } from './runtime/engine';
import { GnosisRegistry } from './runtime/registry';
import { readFileSync, writeFileSync, rmSync } from 'node:fs';
import { resolve, dirname, basename, extname, join } from 'node:path';
import { pathToFileURL } from 'node:url';
import type { ASTNode, ASTEdge, Diagnostic } from './betty/compiler';
import {
  createCertifiedAppEnvelopeFromGgSource,
  createCertifiedAppEnvelopeMirror,
  storeCertifiedAppEnvelopeInQDoc,
  type CertifiedAppEnvelope,
  type CertifiedAppEnvelopeMirror,
} from './certified-app';
import { QDoc } from './crdt/qdoc';
import type { GnosisEngineExecutionCacheState } from './runtime/engine';
import {
  registerTypeScriptBridgeHandlers,
  renderTypeScriptBridgeRuntimeModule,
  type GnosisTypeScriptBridgeBindings,
  type GnosisTypeScriptBridgeResult,
  type GnosisTypeScriptBridgeWave,
} from './ts-bridge';
import type { GnosisExecutionAuthContext } from './auth/core';
import {
  attachRequestScopedReactorHandleGrants,
  generateNativeExecutionReactorGrantSigner,
  resolveNativeExecutionAuth,
  summarizeNativeExecutionAuth,
  type ResolvedNativeExecutionAuth,
  type NativeExecutionAuthSummary,
  type NativeExecutionReactorGrantSigner,
} from './native-execution-auth';
import { FoilEHLARuntime } from './foil-ehla-runtime';
import { ZERO_DEFICIT_FOIL_CERTIFICATE, createStaticFoilAdmissionProof } from './foil-edge-gate';

export type GnodeExecutionStrategy = 'cannon' | 'linear';

export type SerializedGnodeArtifact = Omit<GnosisTypeScriptBridgeResult, 'ast'>;

export interface GnodeEHLAOptions {
  readonly enabled?: boolean;
  readonly maxDebt?: number;
}

export interface GnosisGnodeExecutionSurfaceOptions {
  readonly enabled?: boolean;
  readonly singleThread?: boolean;
  readonly laneCount?: number;
  readonly strategy?: GnodeExecutionStrategy;
  readonly preloadDepth?: number;
  readonly coreCache?: GnosisCoreCache | null;
  readonly runtimeBindingHash?: string | null;
  readonly reactorGrantAudience?: string;
  readonly ehla?: GnodeEHLAOptions;
}

export interface GnodeExecutionControlRequest {
  readonly method: string;
  readonly path?: string;
  readonly query?: string;
  readonly body?: Uint8Array;
  readonly headers?: Record<string, string>;
}

export interface GnodeExecutionControlResponse {
  readonly status: number;
  readonly headers: Record<string, string>;
  readonly body: Uint8Array;
}

export interface RegisterGnodeArtifactRequest {
  readonly compiled: SerializedGnodeArtifact;
  readonly modulePath?: string;
}

export interface ExecuteGnodeArtifactRequest {
  readonly artifactId?: string;
  readonly compiled?: SerializedGnodeArtifact;
  readonly modulePath?: string;
  readonly input?: unknown;
  readonly executionAuth?: GnosisExecutionAuthContext | null;
  readonly executionAuthSource?: ResolvedNativeExecutionAuth['source'];
  readonly laneCount?: number;
  readonly strategy?: GnodeExecutionStrategy;
  readonly ehla?: boolean;
}

export interface GnodeExecutionLanePlan {
  readonly laneIndex: number;
  readonly nodeIds: readonly string[];
}

export interface GnodeExecutionWavePlan {
  readonly index: number;
  readonly kind: GnosisTypeScriptBridgeWave['kind'];
  readonly lanes: readonly GnodeExecutionLanePlan[];
}

export interface GnodeExecutionPlan {
  readonly strategy: GnodeExecutionStrategy;
  readonly laneCount: number;
  readonly preload: 'off' | 'warming' | 'armed';
  readonly waves: readonly GnodeExecutionWavePlan[];
}

export interface GnodeExecutionObservation {
  readonly startedAt: number;
  readonly durationMs: number;
  readonly cacheStatus: GnosisEngineExecutionCacheState['status'] | 'none';
}

export interface GnodeExecutionProfileSnapshot {
  readonly artifactId: string;
  readonly fingerprint: string;
  readonly executions: number;
  readonly cursor: number;
  readonly lastDurationMs: number | null;
  readonly smoothedDurationMs: number | null;
  readonly lastCacheStatus: GnosisEngineExecutionCacheState['status'] | 'none';
  readonly corridorStrength: number;
  readonly observations: readonly GnodeExecutionObservation[];
}

export interface GnodeExecutionEnvelopeStatus {
  readonly artifactId: string;
  readonly envelopeId: string;
  readonly digest: string;
  readonly universeDigest: string;
  readonly appId: string;
  readonly attestationCount: number;
}

export interface GnodeExecutionSurfaceStatus {
  readonly enabled: boolean;
  readonly singleThread: boolean;
  readonly laneCount: number;
  readonly strategy: GnodeExecutionStrategy;
  readonly preloadDepth: number;
  readonly queueDepth: number;
  readonly activeExecutions: number;
  readonly registeredArtifacts: number;
  readonly envelopes: readonly GnodeExecutionEnvelopeStatus[];
  readonly profiles: readonly GnodeExecutionProfileSnapshot[];
  readonly coreCacheEvents: number;
  readonly ehla?: {
    readonly enabled: boolean;
    readonly interpretationDebt: number;
    readonly harvestedTopologies: number;
  };
}

export interface GnodeExecutionResponse {
  readonly artifactId: string;
  readonly fingerprint: string;
  readonly payload: unknown;
  readonly logs: string;
  readonly cache?: GnosisEngineExecutionCacheState;
  readonly plan: GnodeExecutionPlan;
  readonly profile: GnodeExecutionProfileSnapshot;
  readonly envelope: CertifiedAppEnvelopeMirror;
  readonly executionAuth?: NativeExecutionAuthSummary;
  readonly ehla?: {
    readonly mode: 'rigid' | 'harvested';
    readonly debt: number;
  };
}

interface RegisteredArtifact {
  readonly artifactId: string;
  readonly fingerprint: string;
  readonly compiled: SerializedGnodeArtifact;
  readonly modulePath?: string;
  readonly registeredAt: number;
  readonly envelope: CertifiedAppEnvelope;
  /** Hydrated AST — cached to avoid re-parsing GG source on every execution. */
  hydrated?: GnosisTypeScriptBridgeResult;
  /** Cached runtime bindings — avoids re-importing module from disk on every execution. */
  cachedBindings?: GnosisTypeScriptBridgeBindings;
}

interface ProfileState {
  readonly artifactId: string;
  readonly fingerprint: string;
  cursor: number;
  executions: number;
  lastDurationMs: number | null;
  smoothedDurationMs: number | null;
  lastCacheStatus: GnosisEngineExecutionCacheState['status'] | 'none';
  observations: GnodeExecutionObservation[];
}

interface PendingExecution {
  readonly request: ExecuteGnodeArtifactRequest;
  readonly artifact: RegisteredArtifact;
  readonly resolve: (value: GnodeExecutionResponse) => void;
  readonly reject: (reason?: unknown) => void;
}

interface RotationResult<T> {
  readonly lanes: readonly {
    readonly laneIndex: number;
    readonly items: readonly T[];
  }[];
  readonly nextCursor: number;
}

function stableStringify(value: unknown): string {
  if (value === null || value === undefined) {
    return 'null';
  }

  if (typeof value === 'string') {
    return JSON.stringify(value);
  }

  if (typeof value === 'number') {
    return Number.isFinite(value) ? String(value) : 'null';
  }

  if (typeof value === 'boolean') {
    return value ? 'true' : 'false';
  }

  if (Array.isArray(value)) {
    return `[${value.map((item) => stableStringify(item)).join(',')}]`;
  }

  if (typeof value === 'object') {
    const record = value as Record<string, unknown>;
    const entries = Object.entries(record)
      .filter(([, entryValue]) => {
        const entryType = typeof entryValue;
        return (
          entryType !== 'undefined' &&
          entryType !== 'function' &&
          entryType !== 'symbol'
        );
      })
      .sort(([leftKey], [rightKey]) => leftKey.localeCompare(rightKey));

    return `{${entries
      .map(
        ([key, entryValue]) =>
          `${JSON.stringify(key)}:${stableStringify(entryValue)}`
      )
      .join(',')}}`;
  }

  return 'null';
}

function hashArtifact(
  compiled: SerializedGnodeArtifact,
  modulePath?: string
): string {
  return createHash('sha256')
    .update(
      stableStringify({
        exportName: compiled.exportName,
        entryNodeId: compiled.entryNodeId,
        sourceFilePath: compiled.sourceFilePath,
        ggSource: compiled.ggSource,
        topologySource: compiled.topologySource,
        nodePlans: compiled.nodePlans,
        schedule: compiled.schedule,
        modulePath: modulePath ?? null,
      })
    )
    .digest('hex')
    .slice(0, 24);
}

function parseGraphAst(source: string): {
  readonly ast: {
    readonly nodes: Map<string, ASTNode>;
    readonly edges: ASTEdge[];
  };
  readonly diagnostics: readonly Diagnostic[];
} {
  const compiler = new BettyCompiler();
  const parsed = compiler.parse(source);
  const errors = parsed.diagnostics.filter(
    (diagnostic) => diagnostic.severity === 'error'
  );

  if (!parsed.ast || errors.length > 0) {
    const details = errors
      .map(
        (diagnostic) =>
          `${diagnostic.line}:${diagnostic.column} ${diagnostic.message}`
      )
      .join('\n');
    throw new Error(
      details.length > 0
        ? `Failed to compile registered GG artifact:\n${details}`
        : 'Failed to compile registered GG artifact.'
    );
  }

  return {
    ast: parsed.ast,
    diagnostics: parsed.diagnostics,
  };
}

function hydrateArtifact(
  compiled: SerializedGnodeArtifact
): GnosisTypeScriptBridgeResult {
  const parsed = parseGraphAst(compiled.ggSource);
  return {
    ...compiled,
    ast: parsed.ast,
  };
}

function rotateIntoLanes<T>(
  items: readonly T[],
  laneCount: number,
  cursor = 0
): RotationResult<T> {
  const normalizedLaneCount = Math.max(1, laneCount);
  const lanes = Array.from({ length: normalizedLaneCount }, (_, laneIndex) => ({
    laneIndex,
    items: [] as T[],
  }));

  for (let index = 0; index < items.length; index += 1) {
    const laneIndex = (cursor + index) % normalizedLaneCount;
    lanes[laneIndex]!.items.push(items[index]!);
  }

  return {
    lanes,
    nextCursor: (cursor + items.length) % normalizedLaneCount,
  };
}

function jsonResponse(
  status: number,
  body: unknown
): GnodeExecutionControlResponse {
  return {
    status,
    headers: {
      'Content-Type': 'application/json',
      'Cache-Control': 'no-store',
    },
    body: new TextEncoder().encode(JSON.stringify(body, null, 2)),
  };
}

function normalizeRequestPath(request: GnodeExecutionControlRequest): {
  readonly pathname: string;
  readonly search: string;
} {
  const path = request.path ?? '/';
  const queryIndex = path.indexOf('?');
  if (queryIndex >= 0) {
    return {
      pathname: path.slice(0, queryIndex) || '/',
      search: path.slice(queryIndex),
    };
  }

  return {
    pathname: path,
    search: request.query ?? '',
  };
}

function parseJsonBody<T>(request: GnodeExecutionControlRequest): T {
  if (!request.body || request.body.length === 0) {
    throw new Error('Request body is required.');
  }

  return JSON.parse(
    new TextDecoder('utf-8', { fatal: true }).decode(request.body)
  ) as T;
}

function parseInputFromQuery(search: string): unknown {
  const params = new URLSearchParams(
    search.startsWith('?') ? search.slice(1) : search
  );
  const inputJson = params.get('input_json');
  if (!inputJson) {
    return undefined;
  }

  return JSON.parse(inputJson);
}

function snapshotProfile(
  profile: ProfileState,
  corridorStrength: number
): GnodeExecutionProfileSnapshot {
  return {
    artifactId: profile.artifactId,
    fingerprint: profile.fingerprint,
    executions: profile.executions,
    cursor: profile.cursor,
    lastDurationMs: profile.lastDurationMs,
    smoothedDurationMs: profile.smoothedDurationMs,
    lastCacheStatus: profile.lastCacheStatus,
    corridorStrength,
    observations: [...profile.observations],
  };
}

function inferAppId(
  compiled: SerializedGnodeArtifact,
  modulePath?: string
): string {
  const sourcePath = modulePath ?? compiled.sourceFilePath;
  const normalized = basename(sourcePath, extname(sourcePath))
    .replace(/[^A-Za-z0-9._-]+/g, '-')
    .replace(/^-+|-+$/g, '');
  return normalized.length > 0 ? normalized : compiled.exportName;
}

function createEnvelopeStatus(
  artifact: RegisteredArtifact
): GnodeExecutionEnvelopeStatus {
  return {
    artifactId: artifact.artifactId,
    envelopeId: artifact.envelope.envelopeId,
    digest: artifact.envelope.digest,
    universeDigest: artifact.envelope.core.universeDigest,
    appId: artifact.envelope.core.appId,
    attestationCount: artifact.envelope.attestations.length,
  };
}

export class GnosisGnodeExecutionSurface {
  private readonly enabled: boolean;
  private readonly singleThread: boolean;
  private readonly laneCount: number;
  private readonly strategy: GnodeExecutionStrategy;
  private readonly preloadDepth: number;
  private readonly coreCache: GnosisCoreCache;
  private readonly runtimeBindingHash: string | null;
  private readonly reactorGrantAudience: string;
  private readonly reactorGrantSignerPromise: Promise<NativeExecutionReactorGrantSigner>;
  private readonly artifacts = new Map<string, RegisteredArtifact>();
  private readonly profiles = new Map<string, ProfileState>();
  private readonly pendingLanes: PendingExecution[][];
  private readonly ehlaRuntime: FoilEHLARuntime | null;
  private pendingCursor = 0;
  private draining = false;
  private activeExecutions = 0;

  constructor(options: GnosisGnodeExecutionSurfaceOptions = {}) {
    this.enabled = options.enabled ?? false;
    this.singleThread = options.singleThread ?? false;
    this.laneCount = Math.max(1, options.laneCount ?? 4);
    this.strategy = options.strategy ?? 'cannon';
    this.preloadDepth = Math.max(1, options.preloadDepth ?? this.laneCount);
    this.coreCache =
      options.coreCache ??
      new GnosisCoreCache(new QDoc({ guid: 'gnosis-gnode-runtime-cache' }));
    this.runtimeBindingHash = options.runtimeBindingHash?.trim() ?? null;
    this.reactorGrantAudience =
      options.reactorGrantAudience ?? 'aeon://gnosis/native-execution';
    this.reactorGrantSignerPromise = generateNativeExecutionReactorGrantSigner(
      this.runtimeBindingHash ?? 'gnosis-gnode-execution'
    );
    this.pendingLanes = Array.from({ length: this.laneCount }, () => []);
    this.ehlaRuntime = options.ehla?.enabled ? new FoilEHLARuntime({
      maxDebt: options.ehla.maxDebt ?? 10,
      harvestingEnabled: true
    }) : null;
  }

  getStatus(): GnodeExecutionSurfaceStatus {
    const profiles = [...this.profiles.values()]
      .sort((left, right) => left.artifactId.localeCompare(right.artifactId))
      .map((profile) =>
        snapshotProfile(
          profile,
          this.coreCache.corridorStrength(this.corridorKey(profile.fingerprint))
        )
      );

    const envelopes = [...this.artifacts.values()]
      .sort((left, right) => left.artifactId.localeCompare(right.artifactId))
      .map(createEnvelopeStatus);

    const ehlaStats = this.ehlaRuntime?.getEHLAStats();

    return {
      enabled: this.enabled,
      singleThread: this.singleThread,
      laneCount: this.laneCount,
      strategy: this.strategy,
      preloadDepth: this.preloadDepth,
      queueDepth: this.queueDepth(),
      activeExecutions: this.activeExecutions,
      registeredArtifacts: this.artifacts.size,
      envelopes,
      profiles,
      coreCacheEvents: this.coreCache.getEvents().length,
      ...(ehlaStats ? {
        ehla: {
          enabled: true,
          interpretationDebt: ehlaStats.interpretationDebt,
          harvestedTopologies: ehlaStats.harvestedTopologies,
        }
      } : {}),
    };
  }

  registerArtifact(request: RegisterGnodeArtifactRequest): RegisteredArtifact {
    const fingerprint = hashArtifact(request.compiled, request.modulePath);
    const existing = this.artifacts.get(fingerprint);
    if (existing) {
      return existing;
    }

    const envelope = createCertifiedAppEnvelopeFromGgSource({
      appId: inferAppId(request.compiled, request.modulePath),
      sourcePath: request.modulePath ?? request.compiled.sourceFilePath,
      entrypoint: request.compiled.exportName,
      artifactId: fingerprint,
      artifactFingerprint: fingerprint,
      runtimeTarget: 'gnode',
      ggSource: request.compiled.ggSource,
    });
    const artifact: RegisteredArtifact = {
      artifactId: fingerprint,
      fingerprint,
      compiled: request.compiled,
      registeredAt: Date.now(),
      envelope,
      ...(request.modulePath !== undefined
        ? { modulePath: request.modulePath }
        : {}),
    };
    storeCertifiedAppEnvelopeInQDoc(this.coreCache.document, envelope, [
      artifact.artifactId,
      artifact.fingerprint,
      artifact.compiled.sourceFilePath,
      ...(request.modulePath ? [request.modulePath] : []),
    ]);
    this.artifacts.set(artifact.artifactId, artifact);
    return artifact;
  }

  async execute(
    request: ExecuteGnodeArtifactRequest
  ): Promise<GnodeExecutionResponse> {
    const artifact = request.artifactId
      ? this.artifacts.get(request.artifactId)
      : request.compiled
      ? this.registerArtifact({
          compiled: request.compiled,
          ...(request.modulePath !== undefined
            ? { modulePath: request.modulePath }
            : {}),
        })
      : null;

    if (!artifact) {
      throw new Error('Unknown gnode artifact id.');
    }

    if (!this.singleThread) {
      return this.executeTracked(artifact, request);
    }

    return await new Promise<GnodeExecutionResponse>((resolve, reject) => {
      this.enqueue({
        request,
        artifact,
        resolve,
        reject,
      });
      this.pump();
    });
  }

  async handleRequest(
    request: GnodeExecutionControlRequest
  ): Promise<GnodeExecutionControlResponse | null> {
    if (!this.enabled) {
      return null;
    }

    const { pathname, search } = normalizeRequestPath(request);
    if (pathname === '/.aeon/gnode/status' && request.method === 'GET') {
      return jsonResponse(200, this.getStatus());
    }

    if (pathname === '/.aeon/gnode/register' && request.method === 'POST') {
      try {
        const body = parseJsonBody<RegisterGnodeArtifactRequest>(request);
        const artifact = this.registerArtifact(body);
        return jsonResponse(200, {
          artifactId: artifact.artifactId,
          fingerprint: artifact.fingerprint,
          registeredAt: artifact.registeredAt,
          sourceFilePath: artifact.compiled.sourceFilePath,
          envelope: createCertifiedAppEnvelopeMirror(artifact.envelope),
        });
      } catch (error) {
        return jsonResponse(400, {
          error: error instanceof Error ? error.message : String(error),
        });
      }
    }

    if (pathname === '/.aeon/gnode/execute' && request.method === 'POST') {
      try {
        const body = parseJsonBody<ExecuteGnodeArtifactRequest>(request);
        const executionAuthRequest: {
          headers?: Record<string, string>;
          explicitExecutionAuth?: GnosisExecutionAuthContext | null;
          expectedReactorGrantAudience: string;
        } = {
          expectedReactorGrantAudience: this.reactorGrantAudience,
        };
        if (request.headers !== undefined) {
          executionAuthRequest.headers = request.headers;
        }
        if (body.executionAuth !== undefined) {
          executionAuthRequest.explicitExecutionAuth = body.executionAuth;
        }
        const executionAuth = await resolveNativeExecutionAuth(
          executionAuthRequest
        );
        const requestExecutionAuth =
          await this.attachRequestScopedReactorGrants(
            executionAuth.executionAuth,
            request.method,
            pathname
          );
        const result = await this.execute({
          ...body,
          ...(requestExecutionAuth
            ? {
                executionAuth: requestExecutionAuth,
                executionAuthSource: executionAuth.source,
              }
            : {}),
        });
        return jsonResponse(200, result);
      } catch (error) {
        return jsonResponse(400, {
          error: error instanceof Error ? error.message : String(error),
        });
      }
    }

    const executePrefix = '/.aeon/gnode/execute/';
    if (pathname.startsWith(executePrefix) && request.method === 'GET') {
      const artifactId = decodeURIComponent(
        pathname.slice(executePrefix.length)
      );
      try {
        const input = parseInputFromQuery(search);
        const executionAuthRequest: {
          headers?: Record<string, string>;
          expectedReactorGrantAudience: string;
        } = {
          expectedReactorGrantAudience: this.reactorGrantAudience,
        };
        if (request.headers !== undefined) {
          executionAuthRequest.headers = request.headers;
        }
        const executionAuth = await resolveNativeExecutionAuth(
          executionAuthRequest
        );
        const requestExecutionAuth =
          await this.attachRequestScopedReactorGrants(
            executionAuth.executionAuth,
            request.method,
            pathname
          );
        const result = await this.execute({
          artifactId,
          input,
          ...(requestExecutionAuth
            ? {
                executionAuth: requestExecutionAuth,
                executionAuthSource: executionAuth.source,
              }
            : {}),
        });
        return jsonResponse(200, result);
      } catch (error) {
        return jsonResponse(400, {
          error: error instanceof Error ? error.message : String(error),
          artifactId,
        });
      }
    }

    return null;
  }

  /**
   * Load runtime bindings once and cache them on the artifact.
   * Eliminates the per-execution temp file write + dynamic import.
   */
  private async loadAndCacheBindings(
    modulePath: string
  ): Promise<GnosisTypeScriptBridgeBindings> {
    const absolutePath = resolve(modulePath);
    const sourceText = readFileSync(absolutePath, 'utf8');
    const rendered = renderTypeScriptBridgeRuntimeModule(
      sourceText,
      absolutePath
    );

    if (rendered.bindingNames.length === 0) {
      return {};
    }

    // Write temp module, import it, cache the bindings, delete the temp file.
    // This happens exactly ONCE per artifact, not per execution.
    const tempModulePath = join(
      dirname(absolutePath),
      `.${basename(
        absolutePath,
        extname(absolutePath)
      )}.gnode-cached.${Date.now()}.ts`
    );

    writeFileSync(tempModulePath, rendered.moduleSource, 'utf8');
    try {
      const moduleNamespace = (await import(
        `${pathToFileURL(tempModulePath).href}?t=${Date.now()}`
      )) as {
        readonly __gnode_bridge_runtime_bindings?: GnosisTypeScriptBridgeBindings;
      };
      return moduleNamespace.__gnode_bridge_runtime_bindings ?? {};
    } finally {
      rmSync(tempModulePath, { force: true });
    }
  }

  private corridorKey(fingerprint: string): string {
    return `gnosis/gnode/${fingerprint}`;
  }

  private queueDepth(): number {
    return this.pendingLanes.reduce((count, lane) => count + lane.length, 0);
  }

  private async attachRequestScopedReactorGrants(
    executionAuth: GnosisExecutionAuthContext | null,
    method: string,
    path: string
  ): Promise<GnosisExecutionAuthContext | null> {
    if (!executionAuth || !this.runtimeBindingHash) {
      return executionAuth;
    }

    return attachRequestScopedReactorHandleGrants({
      executionAuth,
      runtimeBindingHash: this.runtimeBindingHash,
      audience: this.reactorGrantAudience,
      method,
      path,
      signer: await this.reactorGrantSignerPromise,
    });
  }

  private enqueue(task: PendingExecution): void {
    if (this.strategy === 'linear') {
      this.pendingLanes[0]!.push(task);
      return;
    }

    const rotation = rotateIntoLanes(
      [task],
      this.pendingLanes.length,
      this.pendingCursor
    );
    this.pendingCursor = rotation.nextCursor;
    for (const lane of rotation.lanes) {
      if (lane.items.length === 0) {
        continue;
      }
      this.pendingLanes[lane.laneIndex]!.push(...lane.items);
    }
  }

  private dequeue(): PendingExecution | null {
    if (this.strategy === 'linear') {
      return this.pendingLanes[0]!.shift() ?? null;
    }

    const laneCount = this.pendingLanes.length;
    for (let offset = 0; offset < laneCount; offset += 1) {
      const laneIndex = (this.pendingCursor + offset) % laneCount;
      const task = this.pendingLanes[laneIndex]!.shift();
      if (!task) {
        continue;
      }
      this.pendingCursor = (laneIndex + 1) % laneCount;
      return task;
    }

    return null;
  }

  private pump(): void {
    if (this.draining) {
      return;
    }

    this.draining = true;
    queueMicrotask(async () => {
      try {
        while (true) {
          const task = this.dequeue();
          if (!task) {
            break;
          }

          try {
            task.resolve(
              await this.executeTracked(task.artifact, task.request)
            );
          } catch (error) {
            task.reject(error);
          }
        }
      } finally {
        this.draining = false;
        if (this.queueDepth() > 0) {
          this.pump();
        }
      }
    });
  }

  private getOrCreateProfile(artifact: RegisteredArtifact): ProfileState {
    const existing = this.profiles.get(artifact.fingerprint);
    if (existing) {
      return existing;
    }

    const profile: ProfileState = {
      artifactId: artifact.artifactId,
      fingerprint: artifact.fingerprint,
      cursor: 0,
      executions: 0,
      lastDurationMs: null,
      smoothedDurationMs: null,
      lastCacheStatus: 'none',
      observations: [],
    };
    this.profiles.set(artifact.fingerprint, profile);
    return profile;
  }

  private buildPlan(
    schedule: readonly GnosisTypeScriptBridgeWave[],
    laneCount: number,
    strategy: GnodeExecutionStrategy,
    cursor: number
  ): {
    readonly plan: GnodeExecutionPlan;
    readonly nextCursor: number;
  } {
    const waves: GnodeExecutionWavePlan[] = [];
    let nextCursor = cursor;

    for (const wave of schedule) {
      if (wave.kind === 'parallel') {
        const rotation = rotateIntoLanes(
          wave.nodeIds,
          laneCount,
          strategy === 'cannon' ? nextCursor : 0
        );
        waves.push({
          index: wave.index,
          kind: wave.kind,
          lanes: rotation.lanes
            .filter((lane) => lane.items.length > 0)
            .map((lane) => ({
              laneIndex: lane.laneIndex,
              nodeIds: [...lane.items],
            })),
        });
        if (strategy === 'cannon') {
          nextCursor = rotation.nextCursor;
        }
        continue;
      }

      waves.push({
        index: wave.index,
        kind: wave.kind,
        lanes: [
          {
            laneIndex: 0,
            nodeIds: [...wave.nodeIds],
          },
        ],
      });
    }

    const preload =
      strategy === 'linear'
        ? 'off'
        : this.queueDepth() + this.activeExecutions + 1 >= this.preloadDepth
        ? 'armed'
        : 'warming';

    return {
      plan: {
        strategy,
        laneCount,
        preload,
        waves,
      },
      nextCursor,
    };
  }

  private async executeTracked(
    artifact: RegisteredArtifact,
    request: ExecuteGnodeArtifactRequest
  ): Promise<GnodeExecutionResponse> {
    this.activeExecutions += 1;
    try {
      return await this.executeNow(artifact, request);
    } finally {
      this.activeExecutions -= 1;
    }
  }

  private async executeNow(
    artifact: RegisteredArtifact,
    request: ExecuteGnodeArtifactRequest
  ): Promise<GnodeExecutionResponse> {
    const profile = this.getOrCreateProfile(artifact);
    const laneCount = Math.max(1, request.laneCount ?? this.laneCount);
    const strategy = request.strategy ?? this.strategy;
    const planned = this.buildPlan(
      artifact.compiled.schedule,
      laneCount,
      strategy,
      profile.cursor
    );
    const modulePath =
      request.modulePath ??
      artifact.modulePath ??
      artifact.compiled.sourceFilePath;
    if (modulePath === undefined) {
      throw new Error(
        `Artifact ${artifact.artifactId} is missing a module path for execution`
      );
    }
    const startedAt = Date.now();
    const started = performance.now();
    // Cache hydrated AST to avoid re-parsing GG source on every execution.
    // hydrateArtifact() calls BettyCompiler.parse() which is expensive.
    const mutableArtifact = artifact as {
      hydrated?: GnosisTypeScriptBridgeResult;
      cachedBindings?: GnosisTypeScriptBridgeBindings;
    };
    if (!mutableArtifact.hydrated) {
      mutableArtifact.hydrated = hydrateArtifact(artifact.compiled);
    }

    // Cache runtime bindings to avoid re-importing module from disk on every execution.
    // This eliminates the temp file write + dynamic import that dominated latency.
    if (!mutableArtifact.cachedBindings) {
      mutableArtifact.cachedBindings = await this.loadAndCacheBindings(
        modulePath
      );
    }

    // Fast path: use cached bindings + hydrated AST directly with the engine.
    // Bypasses executeTypeScriptWithGnosis which re-loads bindings every call.
    const registry = new GnosisRegistry();
    registerTypeScriptBridgeHandlers(
      registry,
      mutableArtifact.hydrated,
      mutableArtifact.cachedBindings
    );
    const engine = new GnosisEngine(registry, { coreCache: this.coreCache });

    const ehlaMode = request.ehla !== false && (this.ehlaRuntime !== null || request.ehla === true);
    let ehlaResult: { mode: 'rigid' | 'harvested', debt: number } | undefined;

    if (ehlaMode) {
      const runtime = this.ehlaRuntime ?? new FoilEHLARuntime({ maxDebt: 10 });
      const admission = runtime.admitEHLA({
        ggGraph: new Uint8Array([0]), // Placeholder
        zkProof: createStaticFoilAdmissionProof(ZERO_DEFICIT_FOIL_CERTIFICATE)
      });
      ehlaResult = {
        mode: admission.mode === 'harvested' ? 'harvested' : 'rigid',
        debt: admission.debt
      };
      
      // If we harvested the gap, we can resolve debt in background or just proceed.
      // For first-class support, we'll auto-resolve debt after execution finishes.
    }

    const result = await engine.executeWithResult(
      mutableArtifact.hydrated.ast,
      request.input,
      {
        ...(request.executionAuth
          ? { executionAuth: request.executionAuth }
          : {}),
        cache: {
          corridorKey: this.corridorKey(artifact.fingerprint),
          reuseScope: 'exact',
          suffixKey: request.input ?? null,
        },
      }
    );
    const durationMs = performance.now() - started;

    if (ehlaMode && this.ehlaRuntime) {
      this.ehlaRuntime.resolveDebt(1);
    }

    profile.executions += 1;
    profile.cursor = planned.nextCursor;
    profile.lastDurationMs = durationMs;
    profile.smoothedDurationMs =
      profile.smoothedDurationMs === null
        ? durationMs
        : profile.smoothedDurationMs * 0.7 + durationMs * 0.3;
    profile.lastCacheStatus = result.cache?.status ?? 'none';
    profile.observations.push({
      startedAt,
      durationMs,
      cacheStatus: result.cache?.status ?? 'none',
    });
    if (profile.observations.length > 16) {
      profile.observations.shift();
    }

    const executionAuthSummary = summarizeNativeExecutionAuth(
      request.executionAuth ?? null,
      request.executionAuthSource ?? (request.executionAuth ? 'explicit' : 'none')
    );

    const response: {
      artifactId: string;
      fingerprint: string;
      payload: unknown;
      logs: string;
      plan: GnodeExecutionPlan;
      profile: GnodeExecutionProfileSnapshot;
      envelope: CertifiedAppEnvelopeMirror;
      executionAuth?: NativeExecutionAuthSummary;
      cache?: GnosisEngineExecutionCacheState;
      ehla?: {
        mode: 'rigid' | 'harvested';
        debt: number;
      };
    } = {
      artifactId: artifact.artifactId,
      fingerprint: artifact.fingerprint,
      payload: result.payload,
      logs: result.logs,
      plan: planned.plan,
      profile: snapshotProfile(
        profile,
        this.coreCache.corridorStrength(this.corridorKey(artifact.fingerprint))
      ),
      envelope: createCertifiedAppEnvelopeMirror(artifact.envelope),
    };
    if (executionAuthSummary !== undefined) {
      response.executionAuth = executionAuthSummary;
    }
    if (result.cache !== undefined) {
      response.cache = result.cache;
    }
    if (ehlaResult !== undefined) {
      response.ehla = ehlaResult;
    }
    return response as GnodeExecutionResponse;
  }
}
