// @a0n/gnosis/src/halogram.ts
/**
 * halogram.ts -- System 1 / System 2 via hologram and halogram
 *
 * HOLOGRAM (System 1):
 *   The raw digital twin. No metacognitive optimization. It formalizes the person's
 *   void boundary, playing back their rejection patterns. Fast, automatic,
 *   authentic. What the person would do without thinking about it.
 *
 * HALOGRAM (System 2):
 *   The twin with metacognitive chains (c0-c3) applied at EACH of the
 *   7 personality layers (to attenuate them), then c0-c3 once more on
 *   top of the whole stack. Like a hologram but with self-awareness at
 *   every depth. Slow, deliberate, wise. What the person would do if
 *   they could see all their own patterns.
 *
 *   Layer structure:
 *     Layer 0 (temperament):   c0-c3 attenuates birth configuration
 *     Layer 1 (attachment):    c0-c3 attenuates singularity patterns
 *     Layer 2 (traits):        c0-c3 attenuates complement shape
 *     Layer 3 (behaviors):     c0-c3 attenuates gait selection
 *     Layer 4 (mental health): c0-c3 attenuates kurtosis
 *     Layer 5 (history):       c0-c3 attenuates the boundary itself
 *     Layer 6 (culture):       c0-c3 attenuates inherited void
 *     Layer 7 (META):          c0-c3 on top of the WHOLE stack
 *
 * INTERACTION:
 *   System 1 (hologram) runs first. System 2 (halogram) whispers.
 *   Sometimes System 2 overrides System 1. The override criterion:
 *   if the halogram's complement distribution at the META layer
 *   diverges from the hologram's action by more than a threshold,
 *   System 2 takes over. Otherwise, System 1's fast answer stands.
 *
 *   This is Kahneman's dual process, implemented as void walking.
 */

import {
  type VoidBoundary,
  type Walker,
  type Gait,
  type Measurement,
  type BoundaryStack,
  type TimescaleBoundary,
  type StackWalker,
  createVoidBoundary,
  createWalker,
  createTimescaleBoundary,
  createBoundaryStack,
  createStackWalker,
  createResonance,
  complementDistribution,
  sampleComplement,
  updateVoidBoundary,
  measure,
  measureStack,
  flattenStack,
  stepStackWalker,
  measureStackWalker,
  c0Choose,
  c0Update,
  c1Measure,
  c2c3Adapt,
} from './void';

import {
  type VoidAgent,
  type AgentTick,
  createVoidAgent,
  perceive,
  decide,
  observe,
  reflect,
  adapt,
  tick,
  completeTick,
  personalityVector,
  metacogState,
} from './void-agent';

// ============================================================================
// Hologram -- System 1 (raw digital twin, no metacog optimization)
// ============================================================================

export interface Hologram {
  /** The underlying VoidAgent (no metacog attenuators) */
  agent: VoidAgent;
  /** Label */
  name: string;
}

/**
 * Create a hologram (System 1) from a personality stack.
 * No metacognitive chains -- just the raw void walking.
 */
export function createHologram(
  name: string,
  personality: BoundaryStack,
  actionDimensions: number,
  rng: () => number = Math.random
): Hologram {
  const agent = createVoidAgent(
    {
      name: `${name}:hologram`,
      actionDimensions,
      personality,
      eta: 2.0,
      numHeads: 1, // System 1 is fast -- single head
      neighborhoodRadius: 0, // No spread -- sharp attention
      decayRate: 0.005,
    },
    rng
  );

  return { agent, name };
}

// ============================================================================
// Halogram -- System 2 (metacog at every layer + on top)
// ============================================================================

/**
 * A MetacogChain is a c0-c3 loop attached to a single personality layer.
 * It attenuates the layer: monitors its kurtosis, adapts its eta,
 * controls its exploration rate. The chain formalizes the self-awareness of that layer.
 */
export interface MetacogChain {
  /** Which personality layer this chain attenuates */
  layerIdx: number;
  /** The walker doing the metacognition (walks the layer's void) */
  walker: Walker;
  /** Chain-specific eta (may differ from the layer's walker) */
  eta: number;
  /** Chain-specific exploration */
  exploration: number;
  /** Measurements from this chain's c1 */
  lastMeasurement: Measurement | null;
}

export interface Halogram {
  /** The underlying VoidAgent */
  agent: VoidAgent;
  /** Per-layer metacognitive chains (one per personality layer) */
  layerChains: MetacogChain[];
  /** The META chain -- c0-c3 on top of the whole stack */
  metaChain: MetacogChain;
  /** Override threshold: if META's divergence from hologram exceeds this,
   *  System 2 takes over */
  overrideThreshold: number;
  /** Count of System 2 overrides */
  overrideCount: number;
  /** Count of System 1 passthrough */
  passthroughCount: number;
  /** Label */
  name: string;
}

/**
 * Create a halogram (System 2) from a personality stack.
 * Applies metacognitive chains at each of the 7 personality layers,
 * plus one META chain on top of the whole stack.
 */
export function createHalogram(
  name: string,
  personality: BoundaryStack,
  actionDimensions: number,
  overrideThreshold: number = 0.3,
  rng: () => number = Math.random
): Halogram {
  const agent = createVoidAgent(
    {
      name: `${name}:halogram`,
      actionDimensions,
      personality,
      eta: 2.5, // System 2 starts slightly sharper (more deliberate)
      numHeads: 2, // Multi-head for deeper perception
      neighborhoodRadius: 1,
      decayRate: 0.005,
    },
    rng
  );

  // Per-layer metacog chains
  const layerChains: MetacogChain[] = personality.layers.map((layer, idx) => ({
    layerIdx: idx,
    walker: createWalker(
      createVoidBoundary(layer.boundary.counts.length),
      2.0,
      0.2
    ),
    eta: 2.0,
    exploration: 0.2,
    lastMeasurement: null,
  }));

  // META chain: walks the flattened stack
  const flatDims = personality.layers.reduce(
    (sum, l) => sum + l.boundary.counts.length,
    0
  );
  const metaChain: MetacogChain = {
    layerIdx: -1, // -1 = META (all layers)
    walker: createWalker(createVoidBoundary(flatDims), 2.0, 0.15),
    eta: 2.0,
    exploration: 0.15,
    lastMeasurement: null,
  };

  return {
    agent,
    layerChains,
    metaChain,
    overrideThreshold,
    overrideCount: 0,
    passthroughCount: 0,
    name,
  };
}

/**
 * Attenuate: run c0-c3 on each personality layer via its chain.
 *
 * Each chain's walker monitors its layer's complement distribution.
 * If the layer's kurtosis is too high (rigidity), the chain increases
 * exploration. If too low (chaos), it increases exploitation.
 * This is self-awareness at every depth of personality.
 */
export function attenuate(halogram: Halogram): void {
  const layers = halogram.agent.personality.stack.layers;

  for (const chain of halogram.layerChains) {
    const layer = layers[chain.layerIdx];
    if (!layer) continue;

    // c1: measure the layer
    const m = measure(layer.boundary, chain.eta, chain.walker.steps);
    chain.lastMeasurement = m;

    // c2c3: adapt the chain's parameters based on the layer's kurtosis
    const prevGait = chain.walker.gait;

    // High kurtosis (rigid layer) → increase exploration to soften
    if (m.kurtosis > 1.0) {
      chain.exploration = Math.min(0.5, chain.exploration + 0.02);
      chain.eta = Math.max(1.0, chain.eta - 0.1);
    }
    // Low kurtosis (diffuse layer) → decrease exploration to sharpen
    else if (m.kurtosis < -0.5) {
      chain.exploration = Math.max(0.05, chain.exploration - 0.01);
      chain.eta = Math.min(5.0, chain.eta + 0.05);
    }

    // Apply the chain's attenuated eta back to the layer's walker
    const layerWalker = halogram.agent.personality.walkers[chain.layerIdx];
    if (layerWalker) {
      layerWalker.eta = chain.eta;
      layerWalker.exploration = chain.exploration;
    }

    chain.walker.steps++;
    c2c3Adapt(chain.walker);
  }

  // META chain: measure the whole stack
  const flat = flattenStack(halogram.agent.personality.stack);
  const metaM = measure(
    flat,
    halogram.metaChain.eta,
    halogram.metaChain.walker.steps
  );
  halogram.metaChain.lastMeasurement = metaM;
  halogram.metaChain.walker.steps++;
  c2c3Adapt(halogram.metaChain.walker);
}

// ============================================================================
// Dual Process -- System 1 proposes, System 2 disposes
// ============================================================================

export interface DualProcessResult {
  /** What System 1 (hologram) would do */
  system1Action: number;
  /** What System 2 (halogram) would do */
  system2Action: number;
  /** Which system's action was taken */
  taken: 'system1' | 'system2';
  /** The action that was actually taken */
  action: number;
  /** The divergence that triggered (or didn't trigger) override */
  divergence: number;
  /** System 1 perception weights */
  system1Perception: number[];
  /** System 2 perception weights */
  system2Perception: number[];
  /** META chain measurement */
  metaMeasurement: Measurement | null;
}

/**
 * Run the dual process: System 1 proposes, System 2 whispers.
 *
 * 1. Hologram (System 1) perceives and decides -- fast, automatic
 * 2. Halogram (System 2) attenuates personality layers via metacog chains
 * 3. Halogram perceives and decides -- slow, deliberate
 * 4. Compare: if divergence > threshold, System 2 overrides
 * 5. Otherwise, System 1's answer stands
 *
 * The divergence is measured as the complement weight difference
 * between the two systems' chosen actions. If System 2 strongly
 * disagrees with System 1's choice (low weight in System 2's
 * distribution), it whispers "not that one" and takes over.
 */
export function dualProcess(
  hologram: Hologram,
  halogram: Halogram,
  other?: VoidAgent | null
): DualProcessResult {
  // System 1: fast, automatic
  const s1 = tick(hologram.agent, other);

  // System 2: attenuate first, then perceive
  attenuate(halogram);
  const s2 = tick(halogram.agent, other);

  // Measure divergence: how much does System 2 disagree with System 1's choice?
  const s2Dist = complementDistribution(
    halogram.agent.perception[0].walker.boundary,
    halogram.agent.perception[0].walker.eta
  );
  const s1Dist = complementDistribution(
    hologram.agent.perception[0].walker.boundary,
    hologram.agent.perception[0].walker.eta
  );

  // Divergence: how poorly System 2 rates System 1's choice
  const s2WeightOfS1Choice = s2Dist[s1.action] ?? 0;
  const s2WeightOfS2Choice = s2Dist[s2.action] ?? 0;
  const divergence = Math.abs(s2WeightOfS2Choice - s2WeightOfS1Choice);

  // Override decision
  const override = divergence > halogram.overrideThreshold;
  const taken = override ? 'system2' : 'system1';
  const action = override ? s2.action : s1.action;

  if (override) {
    halogram.overrideCount++;
  } else {
    halogram.passthroughCount++;
  }

  return {
    system1Action: s1.action,
    system2Action: s2.action,
    taken,
    action,
    divergence,
    system1Perception: s1.perception,
    system2Perception: s2.perception,
    metaMeasurement: halogram.metaChain.lastMeasurement,
  };
}

/**
 * Complete a dual process tick after environment feedback.
 * Updates BOTH systems with the outcome.
 */
export function completeDualProcess(
  hologram: Hologram,
  halogram: Halogram,
  result: DualProcessResult,
  rejected: boolean,
  rejectionMagnitude: number = 1,
  reward: number = 0
): void {
  // Update both systems with the actual outcome
  completeTick(
    hologram.agent,
    result.action,
    result.system1Perception,
    rejected,
    rejectionMagnitude,
    reward
  );
  completeTick(
    halogram.agent,
    result.action,
    result.system2Perception,
    rejected,
    rejectionMagnitude,
    reward
  );

  // If System 2 overrode and the outcome was good, reinforce the META chain
  if (result.taken === 'system2' && !rejected) {
    // META chain learns: this override was correct
    c0Update(halogram.metaChain.walker, result.system2Action, 0, reward);
  }
  // If System 2 overrode and the outcome was bad, META chain accumulates void
  if (result.taken === 'system2' && rejected) {
    c0Update(
      halogram.metaChain.walker,
      result.system2Action,
      rejectionMagnitude
    );
  }
}

// ============================================================================
// Introspection
// ============================================================================

export interface DualProcessState {
  system1Gait: Gait;
  system2Gait: Gait;
  metaGait: Gait;
  overrideRate: number;
  layerKurtoses: number[];
  layerEntropies: number[];
  metaEntropy: number;
  metaKurtosis: number;
}

/**
 * Handles the gnosis inspect Dual Process workflow.
 */
export function inspectDualProcess(
  hologram: Hologram,
  halogram: Halogram
): DualProcessState {
  const total = halogram.overrideCount + halogram.passthroughCount;
  return {
    system1Gait: hologram.agent.perception[0].walker.gait,
    system2Gait: halogram.agent.perception[0].walker.gait,
    metaGait: halogram.metaChain.walker.gait,
    overrideRate: total > 0 ? halogram.overrideCount / total : 0,
    layerKurtoses: halogram.layerChains.map(
      (c) => c.lastMeasurement?.kurtosis ?? 0
    ),
    layerEntropies: halogram.layerChains.map(
      (c) => c.lastMeasurement?.entropy ?? 0
    ),
    metaEntropy: halogram.metaChain.lastMeasurement?.entropy ?? 0,
    metaKurtosis: halogram.metaChain.lastMeasurement?.kurtosis ?? 0,
  };
}
