// @a0n/gnosis/src/hiawatha-effect.ts
/**
 * The Hiawatha Effect -- The Long Way Round
 *
 * TypeScript mirror of HiawathaEffect.lean.
 *
 * While not a standard term in academic textbooks, it is often used in
 * design and engineering circles to describe "The Long Way Round."
 * Named after the idea that a straight line is not always the fastest
 * path in a complex landscape -- much like a river (or a cord) that
 * must bend and loop back on itself to eventually reach the sea.
 *
 * Formally: in a topological space with non-zero crossing number, the
 * geodesic requires at least one dilation phase.  A greedy agent that
 * rejects every local increase deadlocks.  The river that refuses to
 * meander never reaches the sea.
 */

// ---------------------------------------------------------------------------
// Part I: The Landscape
// ---------------------------------------------------------------------------

export interface LandscapeNode {
  /** Local topological complexity (crossing number) */
  crossings: number;
  /** Structural elevation (dilation depth) */
  altitude: number;
}

export interface PathStep {
  before: number;
  after: number;
}

/** A step is greedy if it does not increase local complexity. */
export function isGreedyStep(s: PathStep): boolean {
  return s.after <= s.before;
}

/** A step is a meander if it strictly increases local complexity. */
export function isMeander(s: PathStep): boolean {
  return s.after > s.before;
}

// ---------------------------------------------------------------------------
// Part II: Cost Functions
// ---------------------------------------------------------------------------

/** One-time cost of the meander-then-converge strategy. */
export function meanderCost(spike: number, convergenceSteps: number): number {
  return spike + convergenceSteps;
}

/** Cumulative phantom cost of refusing the meander. */
export function refusalCost(deficit: number, rounds: number): number {
  return rounds * deficit;
}

/** Break-even round: the round at which refusal cost exceeds meander cost. */
export function breakEvenRound(spike: number): number {
  // refusal cost = rounds * spike, meander cost = 2 * spike
  // rounds * spike > 2 * spike => rounds > 2
  return 3;
}

// ---------------------------------------------------------------------------
// Part III: River Topology
// ---------------------------------------------------------------------------

export interface RiverSegment {
  /** Euclidean distance to sea */
  straightDistance: number;
  /** Obstacles per unit of straight-line travel */
  terrainResist: number;
}

/** Cost of the straight-line path through terrain. */
export function straightLineCost(r: RiverSegment): number {
  return r.straightDistance * r.terrainResist;
}

/** Cost of the meandering path: longer distance, lower per-unit resistance. */
export function riverMeanderCost(
  r: RiverSegment,
  pathLength: number,
  resistReduction: number
): number {
  return pathLength * (r.terrainResist - resistReduction);
}

/** Does the river meander beat the straight line? */
export function riverMeanderOptimal(
  r: RiverSegment,
  pathLength: number,
  resistReduction: number
): boolean {
  return riverMeanderCost(r, pathLength, resistReduction) < straightLineCost(r);
}

// ---------------------------------------------------------------------------
// Part IV: Cord Topology
// ---------------------------------------------------------------------------

export interface CordPath {
  /** Number of constraint points (eyelets) */
  eyelets: number;
  /** Number of times the cord reverses direction */
  loopbacks: number;
}

/** Apparent waste: each loopback seems redundant. */
export function apparentWaste(c: CordPath): number {
  return c.loopbacks;
}

/** Are all eyelets threaded? loopbacks + 1 >= eyelets */
export function isFullyThreaded(c: CordPath): boolean {
  return c.loopbacks + 1 >= c.eyelets;
}

/** Minimum loopbacks to thread all eyelets. */
export function minimumLoopbacks(eyelets: number): number {
  return Math.max(0, eyelets - 1);
}

// ---------------------------------------------------------------------------
// Part V: Hiawatha Analysis
// ---------------------------------------------------------------------------

export interface HiawathaAnalysis {
  /** Current crossing count */
  crossings: number;
  /** Obstacle height (required spike) */
  obstacle: number;
  /** Does the straight line fail? */
  straightLineFails: boolean;
  /** Is the meander cost bounded? */
  meanderBound: number;
  /** Round at which refusal cost exceeds meander cost */
  refusalBreakEven: number;
  /** Does the geodesic require a meander? */
  geodesicBends: boolean;
  /** Verdict: is the Hiawatha Effect active? */
  hiawathaActive: boolean;
}

/**
 * Analyze a landscape for the Hiawatha Effect.
 *
 * Returns the full diagnostic: whether the straight line fails,
 * the bounded meander cost, the break-even round for refusal,
 * and whether the geodesic must bend.
 */
export function analyzeHiawatha(
  crossings: number,
  obstacle: number
): HiawathaAnalysis {
  const step: PathStep = { before: crossings, after: crossings + obstacle };
  const straightFails = !isGreedyStep(step);
  const bound = meanderCost(obstacle, obstacle); // 2k
  const breakEven = breakEvenRound(obstacle);
  const bends = isMeander(step);
  const active = crossings > 0 && obstacle > 0;

  return {
    crossings,
    obstacle,
    straightLineFails: straightFails,
    meanderBound: bound,
    refusalBreakEven: breakEven,
    geodesicBends: bends,
    hiawathaActive: active,
  };
}

/**
 * Runtime detection: is a system stuck in straight-line refusal?
 *
 * Symptoms: flat progress (no crossing reduction), positive deficit
 * (unresolved obstacle exists), and increasing cumulative cost.
 * Identical signature to the greedy rejection deadlock detector
 * in TopologySampler, but framed as the Hiawatha diagnostic.
 */
export function hiawathaDeadlockDetected(
  recentCrossings: number[],
  deficit: number,
  rounds: number
): boolean {
  if (deficit <= 0 || rounds < 2) return false;
  // Flat: no progress in recent history
  const flat =
    recentCrossings.length >= 2 &&
    recentCrossings.every((c) => c === recentCrossings[0]);
  // Cost exceeds what the meander would have been
  const costExceeded = refusalCost(deficit, rounds) > meanderCost(deficit, deficit);
  return flat && costExceeded;
}
