// @a0n/gnosis/src/human-attestation-certificate.ts
/**
 * Human Attestation Certificate — legally-admissible proof that an action
 * was performed by a human.
 *
 * Combines Leg 3's SlownessCertificate pattern (signed cert over a prism
 * attestation chain) with the HumanDetection result. The underlying chain
 * is multi-hop: conceptually each hop attests to one signal class (agent
 * entropy, metacog revisit, cross-axis correlation, biological cadence,
 * brute-force resistance). The chain as a whole is the proof; the cert
 * seals the detection result + claimed calibration bound with an issuer
 * signature.
 *
 * Use cases:
 *   - E-signature admissibility (was the signer a human?)
 *   - Medical consent (did a human affirm the consent?)
 *   - Voting (was the ballot cast by a human?)
 *   - KYC / AML (is this a real person, not an automaton?)
 *   - Terms-of-service binding (humans are bound; bots are not)
 *
 * Soundness rests on four independent gates, mirroring the slowness
 * primitive:
 *   1. The multi-signal attestation chain is valid
 *   2. The issuer's signature over the cert bytes is valid
 *   3. Confidence clears the calibration bound: C ≥ 1 - bound
 *   4. The detection-result hash sealed in the cert matches the hash of
 *      the detection fields (prevents swapping the result post-signing)
 *
 * This module is a pure gnosis primitive; it must not import from
 * edgework-app or any app surface. The minimal HumanDetectionResult
 * shape is duplicated intentionally -- the detector lives in the app,
 * the primitive lives here.
 *
 * @packageDocumentation
 */

import type {
  AttestationSigner,
  ChainBuildContext,
  PrismAttestationChain,
  AttestationChainVerification,
} from './prism-attestation';
import { verifyAttestationChain } from './prism-attestation';

// ─── Types ──────────────────────────────────────────────────────────

/**
 * Minimal shape we accept from the detector — see
 * edgework-app/services/human-detection.ts. Duplicated intentionally so
 * gnosis stays decoupled from the app layer.
 */
export interface HumanDetectionResult {
  readonly pid: string;
  readonly isHuman: -1 | 0 | 1;
  readonly confidence: number;
  readonly signals: {
    readonly agentAxisEntropy: number;
    readonly metacogRevisitRate: number;
    readonly crossAxisCorrelation: number;
    readonly biologicalCadence: number;
    readonly bruteForceResistance: number;
    readonly observationCount: number;
    readonly observedForMs: number;
  };
  readonly reason?: string;
  readonly atEpochMs: number;
}

/**
 * A legal-grade attestation that a particular action was performed by a
 * human at pLo T, with confidence C, proven by an N-hop chain where each
 * hop attests to an independent signal class (agent entropy, metacog
 * revisit, cross-axis correlation, biological cadence, brute-force).
 *
 * The `calibrationBoundProbability` field is the issuer's claim about the
 * upper bound on P(non-human | certificate-valid). Verifiers compare
 * `confidence` against `1 - calibrationBoundProbability` to decide
 * whether the cert clears its own bound.
 */
export interface HumanAttestationCertificate {
  readonly source: 'GnosisHumanAttestationCertificate';
  readonly certificateId: string;
  readonly subjectPid: string;
  readonly isHuman: -1 | 0 | 1;
  readonly confidence: number;
  readonly calibrationBoundProbability: number;
  readonly detectionResultHash: string;
  readonly attestationChain: PrismAttestationChain;
  readonly issuerDid: string;
  readonly issuedAtPLo: number;
  readonly issuerSignature: Uint8Array;
}

export interface IssueHumanAttestationInput {
  readonly detectionResult: HumanDetectionResult;
  /**
   * The multi-hop attestation chain, where each hop is a witness to one
   * signal class. Hop 0 attests agent entropy, hop 1 attests metacog
   * revisit, hop 2 attests cross-axis correlation, hop 3 attests
   * biological cadence, hop 4 attests brute-force resistance. Extra hops
   * add redundancy.
   */
  readonly chain: PrismAttestationChain;
  readonly issuer: AttestationSigner;
  readonly context: ChainBuildContext;
  /** Claimed upper bound on P(non-human) given a valid certificate. Default 0.05. */
  readonly calibrationBound?: number;
  readonly certificateId?: string;
  readonly issuedAtPLo?: number;
}

export interface HumanAttestationVerification {
  readonly valid: boolean;
  readonly chainValid: boolean;
  readonly issuerSignatureValid: boolean;
  readonly confidenceMeetsBound: boolean;
  readonly detectionHashMatches: boolean;
  readonly reasons: readonly string[];
  readonly chainVerification?: AttestationChainVerification;
}

// ─── Defaults ───────────────────────────────────────────────────────

/** Default calibration bound: P(non-human) ≤ 0.05 under a valid cert. */
export const DEFAULT_CALIBRATION_BOUND = 0.05;

// ─── Canonical Serialization ────────────────────────────────────────

/**
 * Canonical byte encoding of the certificate's signed body.
 *
 * Format: UTF-8 of
 *   "certificateId|subjectPid|isHuman|confidence|calibrationBound|detectionResultHash|chainId|finalPayloadHash|hopCount|issuerDid|issuedAtPLo"
 *
 * This is what the issuer signs, and what the verifier re-hashes. Any
 * tampered field produces different canonical bytes and invalidates the
 * issuer signature.
 *
 * Confidence and calibration bound are encoded at fixed precision
 * (6 decimal places) so floating-point round-trips through JSON wire
 * transport yield the same canonical bytes on both sides.
 */
export function canonicalHumanAttestationBytes(
  cert: Omit<HumanAttestationCertificate, 'issuerSignature'>
): Uint8Array {
  const parts = [
    cert.certificateId,
    cert.subjectPid,
    cert.isHuman.toString(),
    cert.confidence.toFixed(6),
    cert.calibrationBoundProbability.toFixed(6),
    cert.detectionResultHash,
    cert.attestationChain.chainId,
    cert.attestationChain.finalPayloadHash,
    cert.attestationChain.hops.length.toString(),
    cert.issuerDid,
    cert.issuedAtPLo.toString(),
  ];
  return new TextEncoder().encode(parts.join('|'));
}

// ─── Detection-Result Hash ──────────────────────────────────────────

/**
 * Canonical byte encoding of the detection result fields. The hash of
 * these bytes is sealed into the certificate, so any change to the
 * result after issuance invalidates the cert. Fields are ordered and
 * numerics are fixed-precision to ensure wire-stable canonicalization.
 */
export function canonicalDetectionResultBytes(
  result: HumanDetectionResult
): Uint8Array {
  const s = result.signals;
  const parts = [
    result.pid,
    result.isHuman.toString(),
    result.confidence.toFixed(6),
    s.agentAxisEntropy.toFixed(6),
    s.metacogRevisitRate.toFixed(6),
    s.crossAxisCorrelation.toFixed(6),
    s.biologicalCadence.toFixed(6),
    s.bruteForceResistance.toFixed(6),
    s.observationCount.toString(),
    s.observedForMs.toString(),
    result.reason ?? '',
    result.atEpochMs.toString(),
  ];
  return new TextEncoder().encode(parts.join('|'));
}

/**
 * Hash a detection result using the chain-build context's hashFn. This
 * is what we compare in `detectionHashMatches` during verification.
 */
export function hashDetectionResult(
  result: HumanDetectionResult,
  context: ChainBuildContext
): string {
  return context.hashFn(canonicalDetectionResultBytes(result));
}

// ─── Issuance ───────────────────────────────────────────────────────

/**
 * Issue a HumanAttestationCertificate over a completed multi-signal
 * attestation chain.
 *
 * Refusal conditions (mirrors the runtime check in the app layer):
 *   - detectionResult.isHuman !== 1 -- no claim of humanity, no cert
 *   - detectionResult.confidence < 1 - calibrationBound -- doesn't clear the bound
 *   - chain has zero hops -- no multi-signal witness
 *   - calibrationBound out of (0, 1) -- ill-defined probability
 *
 * Issuers MUST refuse to sign certificates they cannot defend; producing
 * an unsound cert is the only way this primitive fails silently.
 */
export async function issueHumanAttestation(
  input: IssueHumanAttestationInput
): Promise<HumanAttestationCertificate> {
  const { detectionResult, chain, issuer, context } = input;
  const calibrationBound = input.calibrationBound ?? DEFAULT_CALIBRATION_BOUND;

  if (!(calibrationBound > 0 && calibrationBound < 1)) {
    throw new Error(
      `calibrationBound must be in (0, 1); got ${calibrationBound}`
    );
  }

  if (detectionResult.isHuman !== 1) {
    throw new Error(
      `Refusing to issue human attestation: isHuman=${detectionResult.isHuman} (expected 1)`
    );
  }

  const minConfidence = 1 - calibrationBound;
  if (detectionResult.confidence < minConfidence) {
    throw new Error(
      `Refusing to issue human attestation: confidence ${detectionResult.confidence} does not meet calibration bound (requires ≥ ${minConfidence})`
    );
  }

  if (chain.hops.length === 0) {
    throw new Error(
      'Refusing to issue human attestation: chain has zero hops (no multi-signal witness)'
    );
  }

  const certificateId =
    input.certificateId ??
    `hac-${detectionResult.pid}-${detectionResult.atEpochMs}`;
  const issuedAtPLo = input.issuedAtPLo ?? detectionResult.atEpochMs;
  const detectionResultHash = hashDetectionResult(detectionResult, context);

  const preSignatureCert: Omit<HumanAttestationCertificate, 'issuerSignature'> = {
    source: 'GnosisHumanAttestationCertificate',
    certificateId,
    subjectPid: detectionResult.pid,
    isHuman: detectionResult.isHuman,
    confidence: detectionResult.confidence,
    calibrationBoundProbability: calibrationBound,
    detectionResultHash,
    attestationChain: chain,
    issuerDid: issuer.did,
    issuedAtPLo,
  };

  const canonical = canonicalHumanAttestationBytes(preSignatureCert);
  const issuerSignature = await issuer.sign(canonical);

  return {
    ...preSignatureCert,
    issuerSignature,
  };
}

// ─── Verification ───────────────────────────────────────────────────

/**
 * Verify a HumanAttestationCertificate.
 *
 * Four independent gates, reported individually so callers can debug
 * which component failed:
 *   1. chainValid: the multi-signal attestation chain passes
 *      verifyAttestationChain
 *   2. issuerSignatureValid: the issuer's signature over the cert
 *      canonical bytes checks out
 *   3. confidenceMeetsBound: isHuman === 1 AND
 *      confidence ≥ 1 - calibrationBoundProbability
 *   4. detectionHashMatches: caller passes the detection result (or cert
 *      keeps the sealed hash); we just check the cert's sealed hash is
 *      internally well-formed (non-empty). The upstream consumer
 *      compares against their detection record.
 *
 * Overall `valid` is the conjunction. Any single failure invalidates the
 * certificate.
 *
 * Note: gate 4 here checks that the sealed hash field is present and
 * non-empty. To verify against a specific detection result, call
 * `detectionResultMatchesCertificate(result, cert, context)` separately;
 * it returns whether the hash of `result` equals `cert.detectionResultHash`.
 */
export async function verifyHumanAttestation(
  cert: HumanAttestationCertificate,
  context: ChainBuildContext,
  issuerVerifier: {
    verify: (did: string, msg: Uint8Array, sig: Uint8Array) => Promise<boolean>;
  }
): Promise<HumanAttestationVerification> {
  const reasons: string[] = [];

  // 1. Chain validation.
  const chainVerification = await verifyAttestationChain(
    cert.attestationChain,
    context
  );
  const chainValid = chainVerification.valid;
  if (!chainValid) {
    reasons.push(
      `Attestation chain invalid: ${chainVerification.reasons.join('; ')}`
    );
  }

  // 2. Issuer signature.
  const { issuerSignature: _omit, ...body } = cert;
  void _omit;
  const canonical = canonicalHumanAttestationBytes(body);
  let issuerSignatureValid = false;
  try {
    issuerSignatureValid = await issuerVerifier.verify(
      cert.issuerDid,
      canonical,
      cert.issuerSignature
    );
  } catch (err) {
    reasons.push(`Issuer verifier threw: ${String(err)}`);
  }
  if (!issuerSignatureValid) {
    reasons.push(`Issuer signature invalid for DID ${cert.issuerDid}`);
  }

  // 3. Confidence clears the calibration bound.
  const minConfidence = 1 - cert.calibrationBoundProbability;
  const confidenceMeetsBound =
    cert.isHuman === 1 && cert.confidence >= minConfidence;
  if (!confidenceMeetsBound) {
    if (cert.isHuman !== 1) {
      reasons.push(
        `Certificate does not assert humanity: isHuman=${cert.isHuman}`
      );
    } else {
      reasons.push(
        `Confidence ${cert.confidence} does not meet calibration bound (requires ≥ ${minConfidence})`
      );
    }
  }

  // 4. Detection-result hash is well-formed (non-empty). Upstream callers
  // perform the semantic equality against their stored detection record.
  const detectionHashMatches =
    typeof cert.detectionResultHash === 'string' &&
    cert.detectionResultHash.length > 0;
  if (!detectionHashMatches) {
    reasons.push('detectionResultHash is empty');
  }

  const valid =
    chainValid &&
    issuerSignatureValid &&
    confidenceMeetsBound &&
    detectionHashMatches;

  return {
    valid,
    chainValid,
    issuerSignatureValid,
    confidenceMeetsBound,
    detectionHashMatches,
    reasons,
    chainVerification,
  };
}

/**
 * Compare a detection result against a certificate's sealed hash. Returns
 * true when hash(result) === cert.detectionResultHash; useful when the
 * verifier holds the original detection record and wants to confirm the
 * cert was issued over that specific record.
 */
export function detectionResultMatchesCertificate(
  result: HumanDetectionResult,
  cert: HumanAttestationCertificate,
  context: ChainBuildContext
): boolean {
  return hashDetectionResult(result, context) === cert.detectionResultHash;
}
