// @a0n/gnosis/src/interlink-topology.ts
/**
 * Interlink Topology Analysis
 *
 * Builds the import/export dependency graph across a codebase and
 * identifies poorly interlinked modules. Low connectivity signals
 * either a true leaf (rare, small utility) or a poorly integrated
 * subsystem that should be better connected.
 *
 * Graph-theoretic measures:
 * - in-degree: how many files import this module
 * - out-degree: how many modules this file imports
 * - betweenness: how often this file is on the shortest path between others
 * - component membership: weakly connected components (islands)
 * - isolation score: files that are large but have few connections
 *
 * Per RepoTopology.lean: crossing_predicts_bugs, beta1_essential.
 * A file with high complexity but low connectivity is a maintenance trap.
 *
 * @module @a0n/gnosis/interlink-topology
 */

import { existsSync, readFileSync, readdirSync, statSync } from 'fs';
import { basename, dirname, extname, join, relative, resolve } from 'path';

// ── Types ────────────────────────────────────────────────────────────

export interface InterlinkNode {
  /** Relative file path from workspace root */
  readonly path: string;
  /** Number of files that import this module */
  readonly inDegree: number;
  /** Number of modules this file imports */
  readonly outDegree: number;
  /** Total degree (in + out) */
  readonly degree: number;
  /** File size in lines */
  readonly lines: number;
  /** Isolation score: lines / max(degree, 1). High = big file, few connections */
  readonly isolationScore: number;
  /** Connected component ID */
  readonly componentId: number;
  /** Whether this is an index/barrel file */
  readonly isBarrel: boolean;
  /** Whether this is a test file */
  readonly isTest: boolean;
}

export interface InterlinkEdge {
  readonly from: string;
  readonly to: string;
  /** Import type: relative path, package, or workspace */
  readonly kind: 'relative' | 'package' | 'workspace';
}

export interface InterlinkComponent {
  readonly id: number;
  readonly size: number;
  readonly files: readonly string[];
  /** Whether this is the main component (largest) */
  readonly isMain: boolean;
}

export interface InterlinkFinding {
  readonly path: string;
  readonly rule: string;
  readonly message: string;
  readonly level: 'warning' | 'hint';
  readonly isolationScore: number;
  readonly inDegree: number;
  readonly outDegree: number;
  readonly lines: number;
  readonly componentId: number;
}

export interface InterlinkAnalysis {
  readonly nodes: readonly InterlinkNode[];
  readonly edges: readonly InterlinkEdge[];
  readonly components: readonly InterlinkComponent[];
  readonly findings: readonly InterlinkFinding[];
  readonly stats: {
    readonly totalFiles: number;
    readonly totalEdges: number;
    readonly totalComponents: number;
    readonly mainComponentSize: number;
    readonly isolatedFiles: number;
    readonly avgDegree: number;
    readonly maxIsolationScore: number;
  };
}

// ── File Discovery ───────────────────────────────────────────────────

const IGNORED_DIRS = new Set([
  'node_modules',
  '.next',
  'dist',
  'build',
  '.git',
  'coverage',
  '.wrangler',
  '__tests__',
  '.a0',
  '.lake',
  '.tmp',
  'storybook-static',
]);

const CODE_EXTENSIONS = new Set([
  '.ts',
  '.tsx',
  '.js',
  '.jsx',
  '.mts',
  '.mjs',
  '.cts',
  '.cjs',
]);

function walkCodeFiles(root: string): string[] {
  const files: string[] = [];
  const queue = [root];

  while (queue.length > 0) {
    const dir = queue.pop()!;
    let entries;
    try {
      entries = readdirSync(dir, { withFileTypes: true });
    } catch {
      continue;
    }

    for (const entry of entries) {
      if (entry.isDirectory()) {
        if (!IGNORED_DIRS.has(entry.name)) queue.push(join(dir, entry.name));
      } else if (entry.isFile() && CODE_EXTENSIONS.has(extname(entry.name))) {
        files.push(join(dir, entry.name));
      }
    }
  }

  return files.sort();
}

// ── Import Extraction ────────────────────────────────────────────────

interface RawImport {
  readonly specifier: string;
  readonly kind: 'relative' | 'package' | 'workspace';
}

function extractImports(content: string): RawImport[] {
  const imports: RawImport[] = [];

  // Static imports: import ... from '...'
  const staticPattern =
    /(?:import|export)\s+(?:[\s\S]*?\s+from\s+)?['"]([^'"]+)['"]/g;
  let match;
  while ((match = staticPattern.exec(content)) !== null) {
    const spec = match[1]!;
    imports.push({ specifier: spec, kind: classifyImport(spec) });
  }

  // Dynamic imports: import('...')  await import('...')
  const dynamicPattern = /import\s*\(\s*['"]([^'"]+)['"]\s*\)/g;
  while ((match = dynamicPattern.exec(content)) !== null) {
    const spec = match[1]!;
    imports.push({ specifier: spec, kind: classifyImport(spec) });
  }

  // require('...')
  const requirePattern = /require\s*\(\s*['"]([^'"]+)['"]\s*\)/g;
  while ((match = requirePattern.exec(content)) !== null) {
    const spec = match[1]!;
    imports.push({ specifier: spec, kind: classifyImport(spec) });
  }

  return imports;
}

function classifyImport(
  specifier: string
): 'relative' | 'package' | 'workspace' {
  if (specifier.startsWith('.') || specifier.startsWith('/')) return 'relative';
  if (
    specifier.startsWith('@a0n/') ||
    specifier.startsWith('@affectively/') ||
    specifier.startsWith('@dashrelay/')
  )
    return 'workspace';
  return 'package';
}

// ── Import Resolution ────────────────────────────────────────────────

const RESOLVE_EXTENSIONS = ['.ts', '.tsx', '.js', '.jsx', '.mts', '.mjs', ''];
const INDEX_FILES = [
  'index.ts',
  'index.tsx',
  'index.js',
  'index.jsx',
  'index.mts',
];

function resolveRelativeImport(
  fromFile: string,
  specifier: string
): string | null {
  const dir = dirname(fromFile);
  const target = resolve(dir, specifier);

  // Try exact match with extensions
  for (const ext of RESOLVE_EXTENSIONS) {
    const candidate = target + ext;
    if (existsSync(candidate)) return candidate;
  }

  // Try as directory with index file
  for (const indexFile of INDEX_FILES) {
    const candidate = join(target, indexFile);
    if (existsSync(candidate)) return candidate;
  }

  return null;
}

// ── Graph Building ───────────────────────────────────────────────────

function isTestFile(path: string): boolean {
  return /\.test\.|\.spec\.|__tests__|\.stories\./.test(path);
}

function isBarrelFile(path: string): boolean {
  const base = basename(path);
  return /^index\.[jt]sx?$/.test(base);
}

function countLines(content: string): number {
  let count = 0;
  for (let i = 0; i < content.length; i++) {
    if (content[i] === '\n') count++;
  }
  return count + 1;
}

/**
 * Builds the Interlink Graph.
 */
export function buildInterlinkGraph(
  root: string,
  dirs?: readonly string[]
): InterlinkAnalysis {
  const allFiles = dirs
    ? dirs.flatMap((d) => walkCodeFiles(resolve(root, d)))
    : walkCodeFiles(root);

  const fileSet = new Set(allFiles);
  const edges: InterlinkEdge[] = [];
  const inDegrees = new Map<string, number>();
  const outDegrees = new Map<string, number>();
  const linesCounts = new Map<string, number>();

  // Initialize all files
  for (const file of allFiles) {
    inDegrees.set(file, 0);
    outDegrees.set(file, 0);
    try {
      linesCounts.set(file, countLines(readFileSync(file, 'utf8')));
    } catch {
      linesCounts.set(file, 0);
    }
  }

  // Extract edges
  for (const file of allFiles) {
    let content;
    try {
      content = readFileSync(file, 'utf8');
    } catch {
      continue;
    }

    const imports = extractImports(content);
    let outCount = 0;

    for (const imp of imports) {
      if (imp.kind !== 'relative') continue; // only track internal edges

      const resolved = resolveRelativeImport(file, imp.specifier);
      if (!resolved || !fileSet.has(resolved)) continue;

      edges.push({
        from: relative(root, file),
        to: relative(root, resolved),
        kind: 'relative',
      });

      inDegrees.set(resolved, (inDegrees.get(resolved) ?? 0) + 1);
      outCount++;
    }

    outDegrees.set(file, outCount);
  }

  // Connected components (union-find)
  const parent = new Map<string, string>();
  const find = (x: string): string => {
    let r = x;
    while (parent.get(r) !== r) r = parent.get(r) ?? r;
    let c = x;
    while (c !== r) {
      const next = parent.get(c) ?? c;
      parent.set(c, r);
      c = next;
    }
    return r;
  };
  const union = (a: string, b: string): void => {
    const ra = find(a);
    const rb = find(b);
    if (ra !== rb) parent.set(ra, rb);
  };

  for (const file of allFiles) parent.set(file, file);
  for (const edge of edges) {
    const fromAbs = resolve(root, edge.from);
    const toAbs = resolve(root, edge.to);
    if (fileSet.has(fromAbs) && fileSet.has(toAbs)) {
      union(fromAbs, toAbs);
    }
  }

  // Assign component IDs
  const componentRoots = new Map<string, number>();
  let nextComponentId = 0;
  const getComponentId = (file: string): number => {
    const r = find(file);
    if (!componentRoots.has(r)) componentRoots.set(r, nextComponentId++);
    return componentRoots.get(r)!;
  };

  // Build nodes
  const nodes: InterlinkNode[] = allFiles.map((file) => {
    const inDeg = inDegrees.get(file) ?? 0;
    const outDeg = outDegrees.get(file) ?? 0;
    const degree = inDeg + outDeg;
    const lines = linesCounts.get(file) ?? 0;

    return {
      path: relative(root, file),
      inDegree: inDeg,
      outDegree: outDeg,
      degree,
      lines,
      isolationScore: lines / Math.max(degree, 1),
      componentId: getComponentId(file),
      isBarrel: isBarrelFile(file),
      isTest: isTestFile(file),
    };
  });

  // Build components
  const componentFiles = new Map<number, string[]>();
  for (const node of nodes) {
    const files = componentFiles.get(node.componentId) ?? [];
    files.push(node.path);
    componentFiles.set(node.componentId, files);
  }

  const mainComponentId =
    [...componentFiles.entries()].sort(
      (a, b) => b[1].length - a[1].length
    )[0]?.[0] ?? 0;

  const components: InterlinkComponent[] = [...componentFiles.entries()]
    .map(([id, files]) => ({
      id,
      size: files.length,
      files,
      isMain: id === mainComponentId,
    }))
    .sort((a, b) => b.size - a.size);

  // Generate findings
  const findings = generateFindings(nodes, components, mainComponentId);

  // Stats
  const totalDegree = nodes.reduce((sum, n) => sum + n.degree, 0);
  const isolatedCount = nodes.filter(
    (n) => n.degree === 0 && !n.isTest && !n.isBarrel && n.lines > 20
  ).length;

  return {
    nodes,
    edges,
    components,
    findings,
    stats: {
      totalFiles: nodes.length,
      totalEdges: edges.length,
      totalComponents: components.length,
      mainComponentSize: components.find((c) => c.isMain)?.size ?? 0,
      isolatedFiles: isolatedCount,
      avgDegree: nodes.length > 0 ? totalDegree / nodes.length : 0,
      maxIsolationScore: Math.max(...nodes.map((n) => n.isolationScore), 0),
    },
  };
}

// ── Finding Generation ───────────────────────────────────────────────

function generateFindings(
  nodes: readonly InterlinkNode[],
  components: readonly InterlinkComponent[],
  mainComponentId: number
): InterlinkFinding[] {
  const findings: InterlinkFinding[] = [];

  for (const node of nodes) {
    if (node.isTest || node.isBarrel) continue;

    // ISLAND: file not in the main component and component has < 5 files
    const comp = components.find((c) => c.id === node.componentId);
    if (comp && !comp.isMain && comp.size < 5 && node.lines > 30) {
      findings.push({
        path: node.path,
        rule: 'TOPOLOGY_ISLAND',
        message: `File is in a disconnected component of ${comp.size} file(s). Either a true leaf or a poorly integrated subsystem.`,
        level: 'warning',
        isolationScore: node.isolationScore,
        inDegree: node.inDegree,
        outDegree: node.outDegree,
        lines: node.lines,
        componentId: node.componentId,
      });
    }

    // DEAD_MODULE: zero in-degree (nobody imports this) and not an entry point
    if (node.inDegree === 0 && node.outDegree > 0 && node.lines > 50) {
      const isEntryPoint =
        /page\.|layout\.|worker\.|main\.|entry\.|index\.|app\.|route\./.test(
          node.path
        );
      if (!isEntryPoint) {
        findings.push({
          path: node.path,
          rule: 'TOPOLOGY_DEAD_MODULE',
          message: `Module has ${node.outDegree} imports but nobody imports it (${node.lines} lines). Possible dead code or missing integration.`,
          level: 'warning',
          isolationScore: node.isolationScore,
          inDegree: 0,
          outDegree: node.outDegree,
          lines: node.lines,
          componentId: node.componentId,
        });
      }
    }

    // LOW_INTEGRATION: large file with very few connections
    if (node.lines > 200 && node.degree <= 2 && node.inDegree <= 1) {
      findings.push({
        path: node.path,
        rule: 'TOPOLOGY_LOW_INTEGRATION',
        message: `${node.lines}-line file with only ${
          node.degree
        } connection(s) (${node.inDegree} in, ${
          node.outDegree
        } out). High isolation score ${Math.round(
          node.isolationScore
        )}. Consider breaking up or integrating more tightly.`,
        level: 'hint',
        isolationScore: node.isolationScore,
        inDegree: node.inDegree,
        outDegree: node.outDegree,
        lines: node.lines,
        componentId: node.componentId,
      });
    }

    // BOTTLENECK: extremely high in-degree relative to average
    if (node.inDegree > 20 && !node.isBarrel) {
      findings.push({
        path: node.path,
        rule: 'TOPOLOGY_BOTTLENECK',
        message: `${node.inDegree} files import this module. High coupling -- changes here ripple widely. Consider splitting responsibilities.`,
        level: 'hint',
        isolationScore: node.isolationScore,
        inDegree: node.inDegree,
        outDegree: node.outDegree,
        lines: node.lines,
        componentId: node.componentId,
      });
    }

    // ONE_WAY: file imports many but nobody imports it back (potential god module)
    if (node.outDegree > 10 && node.inDegree === 0 && node.lines > 100) {
      const isEntryPoint =
        /page\.|layout\.|worker\.|main\.|entry\.|app\.|route\./.test(node.path);
      if (!isEntryPoint) {
        findings.push({
          path: node.path,
          rule: 'TOPOLOGY_ONE_WAY',
          message: `Imports ${node.outDegree} modules but nothing imports it. ${node.lines} lines. One-way dependency -- possible orchestrator that should be an entry point, or dead code.`,
          level: 'warning',
          isolationScore: node.isolationScore,
          inDegree: 0,
          outDegree: node.outDegree,
          lines: node.lines,
          componentId: node.componentId,
        });
      }
    }
  }

  // ORPHAN_COMPONENT: small disconnected components
  for (const comp of components) {
    if (comp.isMain || comp.size >= 5) continue;
    if (comp.files.every((f) => isTestFile(f) || isBarrelFile(f))) continue;

    findings.push({
      path: comp.files[0] ?? '',
      rule: 'TOPOLOGY_ORPHAN_COMPONENT',
      message: `Disconnected component with ${comp.size} file(s): ${comp.files
        .slice(0, 3)
        .join(', ')}${
        comp.size > 3 ? `, +${comp.size - 3} more` : ''
      }. These files have no import path to the main codebase.`,
      level: 'warning',
      isolationScore: 0,
      inDegree: 0,
      outDegree: 0,
      lines: 0,
      componentId: comp.id,
    });
  }

  return findings.sort((a, b) => b.isolationScore - a.isolationScore);
}

// ── Pretty Printing ──────────────────────────────────────────────────

/**
 * Formats the Interlink Report.
 */
export function formatInterlinkReport(analysis: InterlinkAnalysis): string {
  const lines: string[] = [];

  lines.push(
    `Interlink Topology: ${analysis.stats.totalFiles} files, ${analysis.stats.totalEdges} edges, ${analysis.stats.totalComponents} component(s)`
  );
  lines.push(
    `Main component: ${analysis.stats.mainComponentSize} files (${Math.round(
      (analysis.stats.mainComponentSize / analysis.stats.totalFiles) * 100
    )}%)`
  );
  lines.push(
    `Isolated files (>20 lines, zero connections): ${analysis.stats.isolatedFiles}`
  );
  lines.push(`Average degree: ${analysis.stats.avgDegree.toFixed(1)}`);
  lines.push('');

  if (analysis.findings.length === 0) {
    lines.push('No topology findings.');
    return lines.join('\n');
  }

  const byRule = new Map<string, InterlinkFinding[]>();
  for (const f of analysis.findings) {
    const list = byRule.get(f.rule) ?? [];
    list.push(f);
    byRule.set(f.rule, list);
  }

  for (const [rule, items] of [...byRule.entries()].sort(
    (a, b) => b[1].length - a[1].length
  )) {
    lines.push(`${rule} (${items.length}):`);
    for (const f of items.slice(0, 15)) {
      lines.push(`  ${f.path}  ${f.message.slice(0, 100)}`);
    }
    if (items.length > 15) lines.push(`  ...and ${items.length - 15} more`);
    lines.push('');
  }

  return lines.join('\n');
}
