// @a0n/gnosis/src/iota-trees.ts
import type { GraphAST } from './betty/compiler';

export type IotaTree =
  | { readonly kind: 'leaf' }
  | { readonly kind: 'app'; readonly fn: IotaTree; readonly arg: IotaTree };

export interface IotaTopologyNode {
  readonly id: string;
  readonly kind: IotaTree['kind'];
  readonly path: string;
}

export interface IotaTopologyEdge {
  readonly sourceId: string;
  readonly targetId: string;
  readonly role: 'left' | 'right';
}

export interface IotaTopology {
  readonly rootId: string;
  readonly nodes: readonly IotaTopologyNode[];
  readonly edges: readonly IotaTopologyEdge[];
}

export const IOTA_TREE_LEAN_MODULE = 'Gnosis.IotaTrees';

export const iotaLeaf: IotaTree = { kind: 'leaf' };

/**
 * Handles the gnosis iota App workflow.
 */
export function iotaApp(fn: IotaTree, arg: IotaTree): IotaTree {
  return { kind: 'app', fn, arg };
}

function idForPath(path: string): string {
  return `iota_${path.replace(/[^A-Za-z0-9]+/g, '_')}`;
}

function escapeGgString(value: string): string {
  return value.replace(/\\/g, '\\\\').replace(/'/g, "\\'");
}

function collectTopology(
  tree: IotaTree,
  path: string,
  nodes: IotaTopologyNode[],
  edges: IotaTopologyEdge[]
): string {
  const id = idForPath(path);
  nodes.push({ id, kind: tree.kind, path });

  if (tree.kind === 'app') {
    const leftId = collectTopology(tree.fn, `${path}.L`, nodes, edges);
    const rightId = collectTopology(tree.arg, `${path}.R`, nodes, edges);
    edges.push({ sourceId: id, targetId: leftId, role: 'left' });
    edges.push({ sourceId: id, targetId: rightId, role: 'right' });
  }

  return id;
}

/**
 * Handles the gnosis iota Tree To Topology workflow.
 */
export function iotaTreeToTopology(tree: IotaTree): IotaTopology {
  const nodes: IotaTopologyNode[] = [];
  const edges: IotaTopologyEdge[] = [];
  const rootId = collectTopology(tree, 'root', nodes, edges);
  return { rootId, nodes, edges };
}

/**
 * Handles the gnosis iota Tree To Gg workflow.
 */
export function iotaTreeToGg(tree: IotaTree): string {
  const topology = iotaTreeToTopology(tree);
  const lines = [
    `// Iota tree topology; Lean anchor: ${IOTA_TREE_LEAN_MODULE}`,
    `(iota_program:IotaProgram { root: '${topology.rootId}', lean: '${IOTA_TREE_LEAN_MODULE}' })`,
    `(iota_program)-[:ROOT { role: 'iota-root' }]->(${topology.rootId})`,
  ];

  for (const node of topology.nodes) {
    const label = node.kind === 'leaf' ? 'IotaLeaf' : 'IotaApp';
    lines.push(
      `(${node.id}:${label} { kind: '${node.kind}', path: '${escapeGgString(
        node.path
      )}' })`
    );
  }

  for (const edge of topology.edges) {
    const edgeType = edge.role === 'left' ? 'LEFT' : 'RIGHT';
    lines.push(
      `(${edge.sourceId})-[:${edgeType} { role: 'iota-${edge.role}' }]->(${edge.targetId})`
    );
  }

  return `${lines.join('\n')}\n`;
}

function getIotaKind(labels: readonly string[]): IotaTree['kind'] | null {
  if (labels.includes('IotaLeaf')) return 'leaf';
  if (labels.includes('IotaApp')) return 'app';
  return null;
}

function childForRole(
  ast: GraphAST,
  nodeId: string,
  role: IotaTopologyEdge['role']
): string {
  const roleValue = `iota-${role}`;
  const matches = ast.edges.filter(
    (edge) =>
      edge.sourceIds.length === 1 &&
      edge.targetIds.length === 1 &&
      edge.sourceIds[0] === nodeId &&
      edge.properties['role'] === roleValue
  );

  if (matches.length !== 1) {
    throw new Error(
      `Iota topology node '${nodeId}' must have exactly one ${role} edge; found ${matches.length}.`
    );
  }

  const targetId = matches[0]?.targetIds[0];
  if (!targetId) {
    throw new Error(`Iota topology ${role} edge from '${nodeId}' has no target.`);
  }
  return targetId;
}

function iotaChildEdgeCount(ast: GraphAST, nodeId: string): number {
  return ast.edges.filter(
    (edge) =>
      edge.sourceIds.length === 1 &&
      edge.sourceIds[0] === nodeId &&
      (edge.properties['role'] === 'iota-left' ||
        edge.properties['role'] === 'iota-right')
  ).length;
}

function treeFromAstNode(
  ast: GraphAST,
  nodeId: string,
  seen: ReadonlySet<string>
): IotaTree {
  if (seen.has(nodeId)) {
    throw new Error(`Iota topology contains a cycle at '${nodeId}'.`);
  }

  const node = ast.nodes.get(nodeId);
  if (!node) {
    throw new Error(`Iota topology references missing node '${nodeId}'.`);
  }

  const kind = getIotaKind(node.labels);
  if (kind === 'leaf') {
    const childEdgeCount = iotaChildEdgeCount(ast, nodeId);
    if (childEdgeCount !== 0) {
      throw new Error(
        `Iota leaf '${nodeId}' must have no iota children; found ${childEdgeCount}.`
      );
    }
    return iotaLeaf;
  }
  if (kind !== 'app') {
    throw new Error(`Iota topology node '${nodeId}' is not an iota node.`);
  }

  const nextSeen = new Set(seen);
  nextSeen.add(nodeId);
  const left = treeFromAstNode(ast, childForRole(ast, nodeId, 'left'), nextSeen);
  const right = treeFromAstNode(
    ast,
    childForRole(ast, nodeId, 'right'),
    nextSeen
  );
  return iotaApp(left, right);
}

/**
 * Handles the gnosis graph Ast To Iota Tree workflow.
 */
export function graphAstToIotaTree(
  ast: GraphAST,
  rootId = 'iota_root'
): IotaTree {
  return treeFromAstNode(ast, rootId, new Set());
}

/**
 * Handles the gnosis iota Tree Leaf Count workflow.
 */
export function iotaTreeLeafCount(tree: IotaTree): number {
  return tree.kind === 'leaf'
    ? 1
    : iotaTreeLeafCount(tree.fn) + iotaTreeLeafCount(tree.arg);
}

/**
 * Handles the gnosis iota Tree Application Count workflow.
 */
export function iotaTreeApplicationCount(tree: IotaTree): number {
  return tree.kind === 'leaf'
    ? 0
    : iotaTreeApplicationCount(tree.fn) +
        iotaTreeApplicationCount(tree.arg) +
        1;
}

export const pairIotaTree = iotaApp(iotaLeaf, iotaLeaf);

export const leftSpineThreeIotaTree = iotaApp(pairIotaTree, iotaLeaf);
