// @a0n/gnosis/src/juggling-utils.ts
/**
 * Shared numeric and hand-normalization primitives used by both
 * `juggling-orbit-formalization.ts` and `juggling-chaos-orchestration.ts`.
 *
 * Extracted to eliminate duplication across the juggling formal surface.
 */

export type JugglingHand = 'left' | 'right';

/**
 * Modulo that always returns a non-negative value, even for negative dividends.
 *
 * This is crucial for circular calculations (like juggling patterns) where
 * negative remainders would break the mathematical model. JavaScript's % operator
 * returns negative results for negative dividends, which is problematic for
 * periodic systems that require non-negative indices.
 */
export function positiveModulo(value: number, divisor: number): number {
  const remainder = value % divisor;
  return remainder >= 0 ? remainder : remainder + divisor;
}

/**
 * Normalize a number to a positive integer (≥ 1), falling back to `fallback`
 * when the input is undefined or non-finite.
 *
 * This function enforces strict positivity for parameters that cannot be zero
 * or negative, such as counts, speeds, or multipliers. The floor operation
 * ensures we always return an integer, preventing floating-point precision issues
 * in downstream calculations.
 */
export function normalizePositiveInt(
  value: number | undefined,
  fallback: number
): number {
  if (typeof value !== 'number' || !Number.isFinite(value)) {
    return fallback;
  }
  return Math.max(1, Math.floor(value));
}

/**
 * Normalize a number to a non-negative integer (≥ 0), falling back to
 * `fallback` when the input is undefined or non-finite.
 */
export function normalizeNonNegativeInt(
  value: number | undefined,
  fallback: number
): number {
  if (typeof value !== 'number' || !Number.isFinite(value)) {
    return fallback;
  }
  return Math.max(0, Math.floor(value));
}

/**
 * Normalize a number to a finite value, falling back to `fallback` when the
 * input is undefined, NaN, or Infinity.
 */
export function normalizeFinite(
  value: number | undefined,
  fallback: number
): number {
  if (typeof value !== 'number' || !Number.isFinite(value)) {
    return fallback;
  }
  return value;
}

/**
 * Normalize a hand value to canonical 'left' | 'right'.
 * Defaults to 'left' when undefined.
 */
export function normalizeHand(value: JugglingHand | undefined): JugglingHand {
  return value === 'right' ? 'right' : 'left';
}

/**
 * Return the opposite hand.
 *
 * This simple utility is essential for juggling pattern symmetry.
 * Many juggling notations (like siteswap) assume alternating hands,
 * and this function provides the canonical way to compute the mirror
 * hand for any given hand state.
 */
export function oppositeHand(hand: JugglingHand): JugglingHand {
  return hand === 'left' ? 'right' : 'left';
}

/**
 * Round a number to the given number of decimal places.
 *
 * Quantization is essential for deterministic output in geometric
 * calculations and animations. Without quantization, floating-point
 * arithmetic errors accumulate over time, causing visual drift and
 * breaking reproducibility in juggling pattern simulations.
 */
export function quantize(value: number, digits = 4): number {
  const factor = 10 ** digits;
  return Math.round(value * factor) / factor;
}

/**
 * Normalize a speed value to a positive finite number, defaulting to 1.
 *
 * Speed normalization combines safety checks with quantization to ensure
 * juggling animations have stable, predictable timing. The quantization to
 * 3 decimal places prevents micro-variations in speed from causing visual
 * jitter while maintaining sufficient precision for smooth motion.
 */
export function normalizeSpeed(value: number | undefined): number {
  const parsed = normalizeFinite(value, 1);
  return parsed > 0 ? quantize(parsed, 3) : 1;
}

/**
 * Normalize a string label, defaulting to `fallback` when empty/undefined.
 */
export function normalizeLabel(
  value: string | undefined,
  fallback: string
): string {
  const trimmed = value?.trim();
  return trimmed && trimmed.length > 0 ? trimmed : fallback;
}

/**
 * Clamp a number to the unit interval [0, 1].
 *
 * This is fundamental for probability calculations, interpolation,
 * and any normalized parameter space. The explicit branching is
 * faster than Math.min/Math.max for hot paths in animation loops.
 */
export function clamp01(value: number): number {
  if (value < 0) {
    return 0;
  }
  if (value > 1) {
    return 1;
  }
  return value;
}

/**
 * Normalize a value to the unit interval [0, 1], defaulting to `fallback`
 * when undefined or non-finite.
 */
export function normalizeUnitInterval(
  value: number | undefined,
  fallback: number
): number {
  return clamp01(normalizeFinite(value, fallback));
}

/**
 * Return an ISO 8601 timestamp string for the current moment.
 *
 * Provides a standardized, sortable timestamp format for logging
 * and temporal analysis. ISO 8601 ensures timezone-independent
 * representation and lexicographic sorting by chronological order.
 */
export function nowIso(): string {
  return new Date().toISOString();
}
