// @a0n/gnosis/src/juggling-chaos-orchestration.ts
import {
  JUGGLING_ORBIT_THEOREM_IDS,
  deriveJugglingOrbitFormalization,
  type JugglingHand,
  type JugglingOrbitFormalizationBundle,
  type JugglingOrbitThrowSignal,
  type JugglingOrbitTheoremId,
} from './juggling-orbit-formalization';
import {
  positiveModulo,
  normalizePositiveInt,
  normalizeNonNegativeInt,
  normalizeFinite,
  normalizeHand,
  normalizeSpeed,
  normalizeLabel,
  normalizeUnitInterval,
  clamp01,
  quantize as round,
  nowIso,
} from './juggling-utils';

export const JUGGLING_CHAOS_ORCHESTRATION_THEOREM_IDS = [
  'THM-JUGGLING-COMPOSITION-CORPUS-DETERMINISTIC',
  'THM-JUGGLING-COMPOSITION-VALIDITY-GATING',
  'THM-JUGGLING-COMPOSITION-FLOW-NOVELTY-TRADEOFF',
  'THM-JUGGLING-COMPOSITION-EVOLUTION-ELITIST-MONOTONE',
] as const;

export type JugglingChaosOrchestrationTheoremId =
  (typeof JUGGLING_CHAOS_ORCHESTRATION_THEOREM_IDS)[number];

export interface JugglingCompositionSeedPattern {
  readonly patternId: string;
  readonly throws: readonly JugglingOrbitThrowSignal[];
  readonly styleTags?: readonly string[];
}

export interface JugglingCompositionWeights {
  readonly validity: number;
  readonly flow: number;
  readonly novelty: number;
  readonly machineSuitability: number;
}

export const JUGGLING_COMPOSITION_PROPOSERS = [
  'markov',
  'corpus-remix',
  'hybrid',
] as const;

export type JugglingCompositionProposerId =
  (typeof JUGGLING_COMPOSITION_PROPOSERS)[number];

export const JUGGLING_COMPOSITION_PROPOSER_ROUTES = [
  'markov',
  'corpus-remix',
  'elite-carry',
] as const;

export type JugglingCompositionProposerRouteId =
  (typeof JUGGLING_COMPOSITION_PROPOSER_ROUTES)[number];

export interface JugglingGaitProfileInput {
  readonly stepPeriodBeats?: number;
  readonly comFlightTargetBeats?: number;
  readonly comFlightToleranceBeats?: number;
  readonly lateralBias?: number;
  readonly uncertaintySigma?: number;
}

export interface JugglingMpcConfigInput {
  readonly horizonBeats?: number;
  readonly solverBudgetMs?: number;
  readonly restartCount?: number;
  readonly warmStartEnabled?: boolean;
}

export interface JugglingChaosOrchestrationInput {
  readonly period: number;
  readonly corpus: readonly JugglingCompositionSeedPattern[];
  readonly populationSize?: number;
  readonly generations?: number;
  readonly throwsPerPattern?: number;
  readonly proposer?: JugglingCompositionProposerId;
  readonly seed?: number;
  readonly generatedAt?: string;
  readonly source?: string;
  readonly weights?: Partial<JugglingCompositionWeights>;
  readonly gaitProfile?: JugglingGaitProfileInput;
  readonly mpc?: JugglingMpcConfigInput;
  readonly machineOnlyMultiplexThreshold?: number;
  readonly machineOnlyAngleThresholdDeg?: number;
  readonly machineOnlySpeedRatioThreshold?: number;
}

export interface JugglingMpcProxyMetrics {
  readonly horizonPressure: number;
  readonly nonConvexRisk: number;
  readonly dimensionalLoad: number;
  readonly uncertaintyLoad: number;
  readonly warmStartGain: number;
}

export interface JugglingCompositionScore {
  readonly validity: number;
  readonly flow: number;
  readonly gaitCoupling: number;
  readonly novelty: number;
  readonly machineSuitability: number;
  readonly mpc: JugglingMpcProxyMetrics;
  readonly total: number;
  readonly penalties: {
    readonly conservationMismatchRate: number;
    readonly openTransitionRate: number;
  };
}

export interface JugglingCompositionLineage {
  readonly generation: number;
  readonly parentCandidateIds: readonly string[];
  readonly mutationEvents: number;
  readonly seedPatternIds: readonly string[];
  readonly proposerRoute: JugglingCompositionProposerRouteId;
}

export interface JugglingCompositionCandidate {
  readonly candidateId: string;
  readonly signature: string;
  readonly throws: readonly JugglingOrbitThrowSignal[];
  readonly orbit: JugglingOrbitFormalizationBundle;
  readonly score: JugglingCompositionScore;
  readonly lineage: JugglingCompositionLineage;
}

export interface JugglingCompositionGenerationSummary {
  readonly generation: number;
  readonly bestCandidateId: string;
  readonly bestScore: number;
  readonly convergenceSlope: number;
  readonly meanScore: number;
  readonly minScore: number;
  readonly maxScore: number;
  readonly meanMutationEvents: number;
  readonly proposerMix: {
    readonly markov: number;
    readonly corpusRemix: number;
    readonly eliteCarry: number;
  };
  readonly mpc: {
    readonly horizonPressure: number;
    readonly nonConvexRisk: number;
    readonly dimensionalLoad: number;
    readonly uncertaintyLoad: number;
  };
  readonly diversity: number;
}

export interface JugglingChaosOrchestrationBundle {
  readonly generatedAt: string;
  readonly source: string;
  readonly theoremIds: readonly JugglingChaosOrchestrationTheoremId[];
  readonly dependentTheoremIds: readonly JugglingOrbitTheoremId[];
  readonly seed: number;
  readonly period: number;
  readonly configuration: {
    readonly populationSize: number;
    readonly generations: number;
    readonly throwsPerPattern: number;
    readonly proposer: JugglingCompositionProposerId;
    readonly gaitProfile: {
      readonly stepPeriodBeats: number;
      readonly comFlightTargetBeats: number;
      readonly comFlightToleranceBeats: number;
      readonly lateralBias: number;
      readonly uncertaintySigma: number;
    };
    readonly mpc: {
      readonly horizonBeats: number;
      readonly solverBudgetMs: number;
      readonly restartCount: number;
      readonly warmStartEnabled: boolean;
    };
    readonly weights: JugglingCompositionWeights;
  };
  readonly corpusDigest: {
    readonly patternCount: number;
    readonly uniqueThrowTemplateCount: number;
    readonly styleTagCount: number;
  };
  readonly generationSummaries: readonly JugglingCompositionGenerationSummary[];
  readonly winner: JugglingCompositionCandidate;
  readonly finalists: readonly JugglingCompositionCandidate[];
}

interface ThrowTemplate {
  readonly fromHand: JugglingHand;
  readonly toHand?: JugglingHand;
  readonly flightBeats: number;
  readonly lane: number;
  readonly angleDeg: number;
  readonly speed: number;
  readonly label: string;
}

interface SeedPatternModel {
  readonly patternId: string;
  readonly throws: readonly JugglingOrbitThrowSignal[];
  readonly templateSignatures: readonly string[];
  readonly styleTags: readonly string[];
}

interface CompositionModel {
  readonly templates: readonly ThrowTemplate[];
  readonly transitions: ReadonlyMap<string, readonly ThrowTemplate[]>;
  readonly seedPatterns: readonly SeedPatternModel[];
  readonly uniqueTemplateCount: number;
  readonly styleTagCount: number;
}

interface NormalizedGaitProfile {
  readonly stepPeriodBeats: number;
  readonly comFlightTargetBeats: number;
  readonly comFlightToleranceBeats: number;
  readonly lateralBias: number;
  readonly uncertaintySigma: number;
}

interface NormalizedMpcConfig {
  readonly horizonBeats: number;
  readonly solverBudgetMs: number;
  readonly restartCount: number;
  readonly warmStartEnabled: boolean;
}

interface Genome {
  readonly throws: readonly JugglingOrbitThrowSignal[];
  readonly parentCandidateIds: readonly string[];
  readonly seedPatternIds: readonly string[];
  readonly mutationEvents: number;
  readonly proposerRoute: JugglingCompositionProposerRouteId;
}

const DEFAULT_WEIGHTS: JugglingCompositionWeights = {
  validity: 0.42,
  flow: 0.33,
  novelty: 0.17,
  machineSuitability: 0.08,
};

const DEFAULT_SEED = 20260411;
const DEFAULT_POPULATION_SIZE = 24;
const DEFAULT_GENERATIONS = 8;
const DEFAULT_THROWS_PER_PATTERN = 12;
const DEFAULT_PROPOSER: JugglingCompositionProposerId = 'markov';
const DEFAULT_GAIT_PROFILE: NormalizedGaitProfile = {
  stepPeriodBeats: 2,
  comFlightTargetBeats: 2.7,
  comFlightToleranceBeats: 1.3,
  lateralBias: 0.55,
  uncertaintySigma: 0.18,
};
const DEFAULT_MPC_CONFIG: NormalizedMpcConfig = {
  horizonBeats: 6,
  solverBudgetMs: 8,
  restartCount: 2,
  warmStartEnabled: true,
};

// Utility functions imported from juggling-utils.ts

function createRng(seed: number): () => number {
  let state = seed >>> 0;
  if (state === 0) {
    state = 0x6d2b79f5;
  }

  return () => {
    state = (state + 0x6d2b79f5) >>> 0;
    let t = Math.imul(state ^ (state >>> 15), 1 | state);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

function normalizeProposerId(
  proposer: JugglingCompositionProposerId | undefined
): JugglingCompositionProposerId {
  if (!proposer) {
    return DEFAULT_PROPOSER;
  }
  return JUGGLING_COMPOSITION_PROPOSERS.includes(proposer)
    ? proposer
    : DEFAULT_PROPOSER;
}

function normalizeGaitProfile(
  gaitProfile: JugglingGaitProfileInput | undefined
): NormalizedGaitProfile {
  return {
    stepPeriodBeats: normalizePositiveInt(
      gaitProfile?.stepPeriodBeats,
      DEFAULT_GAIT_PROFILE.stepPeriodBeats
    ),
    comFlightTargetBeats: Math.max(
      1,
      normalizeFinite(
        gaitProfile?.comFlightTargetBeats,
        DEFAULT_GAIT_PROFILE.comFlightTargetBeats
      )
    ),
    comFlightToleranceBeats: Math.max(
      0.25,
      normalizeFinite(
        gaitProfile?.comFlightToleranceBeats,
        DEFAULT_GAIT_PROFILE.comFlightToleranceBeats
      )
    ),
    lateralBias: normalizeUnitInterval(
      gaitProfile?.lateralBias,
      DEFAULT_GAIT_PROFILE.lateralBias
    ),
    uncertaintySigma: normalizeUnitInterval(
      gaitProfile?.uncertaintySigma,
      DEFAULT_GAIT_PROFILE.uncertaintySigma
    ),
  };
}

function normalizeMpcConfig(
  mpcConfig: JugglingMpcConfigInput | undefined
): NormalizedMpcConfig {
  const warmStartEnabled =
    typeof mpcConfig?.warmStartEnabled === 'boolean'
      ? mpcConfig.warmStartEnabled
      : DEFAULT_MPC_CONFIG.warmStartEnabled;
  return {
    horizonBeats: normalizePositiveInt(
      mpcConfig?.horizonBeats,
      DEFAULT_MPC_CONFIG.horizonBeats
    ),
    solverBudgetMs: Math.max(
      1,
      normalizeFinite(mpcConfig?.solverBudgetMs, DEFAULT_MPC_CONFIG.solverBudgetMs)
    ),
    restartCount: normalizeNonNegativeInt(
      mpcConfig?.restartCount,
      DEFAULT_MPC_CONFIG.restartCount
    ),
    warmStartEnabled,
  };
}

function pickOne<T>(values: readonly T[], rng: () => number): T | undefined {
  if (values.length === 0) {
    return undefined;
  }
  const index = Math.floor(rng() * values.length);
  return values[index];
}

function templateSignature(template: ThrowTemplate): string {
  return [
    `f${template.fromHand}`,
    `t${template.toHand ?? '-'}`,
    `h${template.flightBeats}`,
    `l${template.lane}`,
    `a${template.angleDeg.toFixed(1)}`,
    `s${template.speed.toFixed(2)}`,
  ].join(':');
}

function toTemplate(
  throwSignal: JugglingOrbitThrowSignal,
  fallbackLabel: string
): ThrowTemplate {
  return {
    fromHand: normalizeHand(throwSignal.fromHand),
    ...(throwSignal.toHand !== undefined
      ? { toHand: normalizeHand(throwSignal.toHand) }
      : {}),
    flightBeats: normalizePositiveInt(throwSignal.flightBeats, 1),
    lane: normalizeNonNegativeInt(throwSignal.lane, 0),
    angleDeg: round(normalizeFinite(throwSignal.angleDeg, 0), 1),
    speed: normalizeSpeed(throwSignal.speed),
    label: normalizeLabel(throwSignal.label, fallbackLabel),
  } satisfies ThrowTemplate;
}

function toThrowSignal(
  template: ThrowTemplate,
  beat: number,
  labelSuffix: string
): JugglingOrbitThrowSignal {
  return {
    beat,
    fromHand: template.fromHand,
    ...(template.toHand !== undefined ? { toHand: template.toHand } : {}),
    flightBeats: template.flightBeats,
    lane: template.lane,
    angleDeg: template.angleDeg,
    speed: template.speed,
    label: `${template.label}-${labelSuffix}`,
  } satisfies JugglingOrbitThrowSignal;
}

function cloneThrows(
  throwsSignals: readonly JugglingOrbitThrowSignal[]
): readonly JugglingOrbitThrowSignal[] {
  return throwsSignals.map((throwSignal) => ({ ...throwSignal }));
}

function normalizeThrowSequence(
  throwsSignals: readonly JugglingOrbitThrowSignal[],
  period: number
): readonly JugglingOrbitThrowSignal[] {
  return throwsSignals.map((throwSignal, index) => {
    const template = toTemplate(throwSignal, `throw-${index + 1}`);
    return {
      beat: positiveModulo(index, period),
      fromHand: template.fromHand,
      ...(template.toHand !== undefined ? { toHand: template.toHand } : {}),
      flightBeats: template.flightBeats,
      lane: template.lane,
      angleDeg: template.angleDeg,
      speed: template.speed,
      label: template.label,
    } satisfies JugglingOrbitThrowSignal;
  });
}

function buildModel(
  corpus: readonly JugglingCompositionSeedPattern[],
  period: number
): CompositionModel {
  const templates: ThrowTemplate[] = [];
  const transitionBuilder = new Map<string, ThrowTemplate[]>();
  const seedPatterns: SeedPatternModel[] = [];
  const templateSignatures = new Set<string>();
  const styleTags = new Set<string>();

  for (const pattern of corpus) {
    const normalizedThrows = normalizeThrowSequence(pattern.throws, period);
    if (normalizedThrows.length === 0) {
      continue;
    }

    const normalizedPatternId = normalizeLabel(
      pattern.patternId,
      `pattern-${seedPatterns.length + 1}`
    );
    const throwTemplates = normalizedThrows.map((throwSignal, index) =>
      toTemplate(throwSignal, `${normalizedPatternId}-t${index + 1}`)
    );
    const signatures = throwTemplates.map(templateSignature);

    seedPatterns.push({
      patternId: normalizedPatternId,
      throws: normalizedThrows,
      templateSignatures: [...new Set(signatures)],
      styleTags: pattern.styleTags ? [...pattern.styleTags] : [],
    });

    for (const styleTag of pattern.styleTags ?? []) {
      styleTags.add(styleTag);
    }

    for (const template of throwTemplates) {
      templates.push(template);
      templateSignatures.add(templateSignature(template));
    }

    for (let index = 0; index < throwTemplates.length; index += 1) {
      const current = throwTemplates[index];
      const next = throwTemplates[(index + 1) % throwTemplates.length];
      if (!current || !next) {
        continue;
      }
      const key = templateSignature(current);
      const existing = transitionBuilder.get(key);
      if (existing) {
        existing.push(next);
      } else {
        transitionBuilder.set(key, [next]);
      }
    }
  }

  if (templates.length === 0) {
    const fallback: ThrowTemplate[] = [
      {
        fromHand: 'left',
        toHand: 'right',
        flightBeats: 3,
        lane: 0,
        angleDeg: 10,
        speed: 7,
        label: 'fallback-a',
      },
      {
        fromHand: 'right',
        toHand: 'left',
        flightBeats: 3,
        lane: 0,
        angleDeg: 10,
        speed: 7,
        label: 'fallback-b',
      },
    ];
    for (let index = 0; index < fallback.length; index += 1) {
      const current = fallback[index];
      const next = fallback[(index + 1) % fallback.length];
      if (!current || !next) {
        continue;
      }
      templates.push(current);
      templateSignatures.add(templateSignature(current));
      transitionBuilder.set(templateSignature(current), [next]);
    }
    seedPatterns.push({
      patternId: 'fallback',
      throws: normalizeThrowSequence(
        fallback.map((template, index) =>
          toThrowSignal(template, index % period, String(index + 1))
        ),
        period
      ),
      templateSignatures: [...templateSignatures],
      styleTags: [],
    });
  }

  const transitions = new Map<string, readonly ThrowTemplate[]>();
  for (const [key, values] of transitionBuilder.entries()) {
    transitions.set(key, values);
  }

  seedPatterns.sort((left, right) => left.patternId.localeCompare(right.patternId));

  return {
    templates,
    transitions,
    seedPatterns,
    uniqueTemplateCount: templateSignatures.size,
    styleTagCount: styleTags.size,
  };
}

function composePattern(
  model: CompositionModel,
  period: number,
  throwsPerPattern: number,
  rng: () => number
): readonly JugglingOrbitThrowSignal[] {
  const output: JugglingOrbitThrowSignal[] = [];
  let cursor =
    pickOne(model.templates, rng) ??
    model.templates[0] ?? {
      fromHand: 'left',
      toHand: 'right',
      flightBeats: 3,
      lane: 0,
      angleDeg: 10,
      speed: 7,
      label: 'fallback',
    };

  for (let index = 0; index < throwsPerPattern; index += 1) {
    const beat = positiveModulo(index, period);
    output.push(toThrowSignal(cursor, beat, String(index + 1)));

    const next =
      pickOne(model.transitions.get(templateSignature(cursor)) ?? [], rng) ??
      pickOne(model.templates, rng);
    if (next) {
      cursor = next;
    }
  }

  return normalizeThrowSequence(output, period);
}

function fitPatternLength(
  throwsSignals: readonly JugglingOrbitThrowSignal[],
  targetLength: number,
  rng: () => number
): readonly JugglingOrbitThrowSignal[] {
  const normalizedTarget = Math.max(1, targetLength);
  if (throwsSignals.length === normalizedTarget) {
    return cloneThrows(throwsSignals);
  }
  if (throwsSignals.length > normalizedTarget) {
    return cloneThrows(throwsSignals.slice(0, normalizedTarget));
  }
  if (throwsSignals.length === 0) {
    return [];
  }

  const expanded = [...throwsSignals.map((throwSignal) => ({ ...throwSignal }))];
  while (expanded.length < normalizedTarget) {
    const pick = pickOne(throwsSignals, rng) ?? throwsSignals[expanded.length % throwsSignals.length];
    if (!pick) {
      break;
    }
    expanded.push({ ...pick });
  }
  return expanded;
}

function remixFromSeedPatterns(
  model: CompositionModel,
  period: number,
  throwsPerPattern: number,
  rng: () => number
): readonly JugglingOrbitThrowSignal[] {
  const output: JugglingOrbitThrowSignal[] = [];
  if (model.seedPatterns.length === 0) {
    return composePattern(model, period, throwsPerPattern, rng);
  }

  while (output.length < throwsPerPattern) {
    const seedPattern = pickOne(model.seedPatterns, rng) ?? model.seedPatterns[0];
    if (!seedPattern || seedPattern.throws.length === 0) {
      break;
    }
    const chunkLength = Math.min(
      seedPattern.throws.length,
      Math.max(1, 1 + Math.floor(rng() * 4))
    );
    const startIndex = Math.floor(rng() * seedPattern.throws.length);
    for (let offset = 0; offset < chunkLength && output.length < throwsPerPattern; offset += 1) {
      const source =
        seedPattern.throws[(startIndex + offset) % seedPattern.throws.length] ??
        seedPattern.throws[0];
      if (!source) {
        continue;
      }
      output.push({ ...source });
    }
  }

  if (output.length === 0) {
    return composePattern(model, period, throwsPerPattern, rng);
  }
  return normalizeThrowSequence(output.slice(0, throwsPerPattern), period);
}

function crossoverPattern(
  left: readonly JugglingOrbitThrowSignal[],
  right: readonly JugglingOrbitThrowSignal[],
  period: number,
  rng: () => number
): readonly JugglingOrbitThrowSignal[] {
  const minLength = Math.min(left.length, right.length);
  if (minLength < 2) {
    return normalizeThrowSequence(cloneThrows(left), period);
  }

  const split = 1 + Math.floor(rng() * (minLength - 1));
  const merged = [...left.slice(0, split), ...right.slice(split)];
  return normalizeThrowSequence(merged, period);
}

function remixParentsAndCorpus(
  parentA: readonly JugglingOrbitThrowSignal[],
  parentB: readonly JugglingOrbitThrowSignal[],
  model: CompositionModel,
  period: number,
  throwsPerPattern: number,
  rng: () => number
): readonly JugglingOrbitThrowSignal[] {
  const base = fitPatternLength(parentA, throwsPerPattern, rng).map((throwSignal) => ({
    ...throwSignal,
  }));
  const donor = fitPatternLength(parentB, throwsPerPattern, rng);

  if (base.length === 0) {
    return remixFromSeedPatterns(model, period, throwsPerPattern, rng);
  }

  if (donor.length > 0) {
    const insertStart = Math.floor(rng() * base.length);
    const donorStart = Math.floor(rng() * donor.length);
    const segmentLength = Math.max(
      1,
      Math.min(4, donor.length, base.length - insertStart)
    );
    for (let offset = 0; offset < segmentLength; offset += 1) {
      const source = donor[(donorStart + offset) % donor.length];
      if (!source) {
        continue;
      }
      base[insertStart + offset] = { ...source };
    }
  }

  const seedPattern = pickOne(model.seedPatterns, rng);
  if (seedPattern && seedPattern.throws.length > 0 && base.length > 0) {
    const insertStart = Math.floor(rng() * base.length);
    const seedStart = Math.floor(rng() * seedPattern.throws.length);
    const segmentLength = Math.max(
      1,
      Math.min(3, seedPattern.throws.length, base.length - insertStart)
    );
    for (let offset = 0; offset < segmentLength; offset += 1) {
      const source =
        seedPattern.throws[(seedStart + offset) % seedPattern.throws.length];
      if (!source) {
        continue;
      }
      base[insertStart + offset] = { ...source };
    }
  }

  return normalizeThrowSequence(base, period);
}

interface ProposedPattern {
  readonly throws: readonly JugglingOrbitThrowSignal[];
  readonly proposerRoute: JugglingCompositionProposerRouteId;
}

function proposeInitialPattern(
  proposer: JugglingCompositionProposerId,
  model: CompositionModel,
  period: number,
  throwsPerPattern: number,
  rng: () => number
): ProposedPattern {
  if (proposer === 'markov') {
    return {
      throws: composePattern(model, period, throwsPerPattern, rng),
      proposerRoute: 'markov',
    };
  }
  if (proposer === 'corpus-remix') {
    return {
      throws: remixFromSeedPatterns(model, period, throwsPerPattern, rng),
      proposerRoute: 'corpus-remix',
    };
  }
  return rng() < 0.5
    ? {
        throws: composePattern(model, period, throwsPerPattern, rng),
        proposerRoute: 'markov',
      }
    : {
        throws: remixFromSeedPatterns(model, period, throwsPerPattern, rng),
        proposerRoute: 'corpus-remix',
      };
}

function proposeOffspringPattern(
  proposer: JugglingCompositionProposerId,
  parentA: JugglingCompositionCandidate,
  parentB: JugglingCompositionCandidate,
  model: CompositionModel,
  period: number,
  throwsPerPattern: number,
  rng: () => number
): ProposedPattern {
  if (proposer === 'markov') {
    return {
      throws:
        rng() < 0.72
          ? crossoverPattern(parentA.throws, parentB.throws, period, rng)
          : fitPatternLength(parentA.throws, throwsPerPattern, rng),
      proposerRoute: 'markov',
    };
  }
  if (proposer === 'corpus-remix') {
    return {
      throws: remixParentsAndCorpus(
        parentA.throws,
        parentB.throws,
        model,
        period,
        throwsPerPattern,
        rng
      ),
      proposerRoute: 'corpus-remix',
    };
  }
  return rng() < 0.55
    ? {
        throws: crossoverPattern(parentA.throws, parentB.throws, period, rng),
        proposerRoute: 'markov',
      }
    : {
        throws: remixParentsAndCorpus(
          parentA.throws,
          parentB.throws,
          model,
          period,
          throwsPerPattern,
          rng
        ),
        proposerRoute: 'corpus-remix',
      };
}

function mutatePattern(
  throwsSignals: readonly JugglingOrbitThrowSignal[],
  period: number,
  rng: () => number
): { throws: readonly JugglingOrbitThrowSignal[]; mutationEvents: number } {
  if (throwsSignals.length === 0) {
    return { throws: throwsSignals, mutationEvents: 0 };
  }

  const mutable = throwsSignals.map((throwSignal) => ({ ...throwSignal }));
  const eventCount = 1 + Math.floor(rng() * Math.min(3, mutable.length));
  let mutationEvents = 0;

  for (let event = 0; event < eventCount; event += 1) {
    const index = Math.floor(rng() * mutable.length);
    const current = mutable[index];
    if (!current) {
      continue;
    }

    const op = Math.floor(rng() * 6);
    if (op === 0) {
      current.angleDeg = round(
        normalizeFinite(current.angleDeg, 0) + (rng() - 0.5) * 12,
        1
      );
    } else if (op === 1) {
      current.speed = round(
        Math.max(1, normalizeSpeed(current.speed) * (0.85 + rng() * 0.35)),
        3
      );
    } else if (op === 2) {
      current.lane = Math.max(
        0,
        normalizeNonNegativeInt(current.lane, 0) + (rng() < 0.5 ? -1 : 1)
      );
    } else if (op === 3) {
      current.flightBeats = Math.max(
        1,
        normalizePositiveInt(current.flightBeats, 1) + (rng() < 0.5 ? -1 : 1)
      );
    } else if (op === 4) {
      current.toHand = normalizeHand(current.toHand) === 'left' ? 'right' : 'left';
    } else {
      current.fromHand = current.fromHand === 'left' ? 'right' : 'left';
    }

    mutationEvents += 1;
  }

  return {
    throws: normalizeThrowSequence(mutable, period),
    mutationEvents,
  };
}

function mean(values: readonly number[]): number {
  if (values.length === 0) {
    return 0;
  }
  return values.reduce((sum, value) => sum + value, 0) / values.length;
}

function deriveFlowScore(
  throwsSignals: readonly JugglingOrbitThrowSignal[]
): number {
  if (throwsSignals.length < 2) {
    return 0.5;
  }

  let handSwitches = 0;
  const angleDelta: number[] = [];
  const speedDelta: number[] = [];
  const laneDelta: number[] = [];

  for (let index = 1; index < throwsSignals.length; index += 1) {
    const previous = throwsSignals[index - 1];
    const current = throwsSignals[index];
    if (!previous || !current) {
      continue;
    }

    if (normalizeHand(previous.fromHand) !== normalizeHand(current.fromHand)) {
      handSwitches += 1;
    }

    angleDelta.push(
      Math.abs(
        normalizeFinite(current.angleDeg, 0) - normalizeFinite(previous.angleDeg, 0)
      )
    );
    speedDelta.push(
      Math.abs(
        normalizeSpeed(current.speed) - normalizeSpeed(previous.speed)
      )
    );
    laneDelta.push(
      Math.abs(
        normalizeNonNegativeInt(current.lane, 0) -
          normalizeNonNegativeInt(previous.lane, 0)
      )
    );
  }

  const transitions = Math.max(1, throwsSignals.length - 1);
  const alternationRatio = handSwitches / transitions;
  const alternationFit = clamp01(1 - Math.abs(alternationRatio - 0.55) / 0.55);
  const angleSmoothness = clamp01(1 - mean(angleDelta) / 22);
  const speedSmoothness = clamp01(1 - mean(speedDelta) / 4.5);
  const laneSmoothness = clamp01(1 - mean(laneDelta) / 2.25);

  return round(
    0.34 * angleSmoothness +
      0.28 * speedSmoothness +
      0.2 * laneSmoothness +
      0.18 * alternationFit
  );
}

function deriveGaitCouplingScore(
  throwsSignals: readonly JugglingOrbitThrowSignal[],
  gaitProfile: NormalizedGaitProfile
): number {
  if (throwsSignals.length === 0) {
    return 0.5;
  }

  const flightCloseness = mean(
    throwsSignals.map((throwSignal) => {
      const delta = Math.abs(throwSignal.flightBeats - gaitProfile.comFlightTargetBeats);
      return clamp01(1 - delta / gaitProfile.comFlightToleranceBeats);
    })
  );

  let handSwitches = 0;
  for (let index = 1; index < throwsSignals.length; index += 1) {
    const previous = throwsSignals[index - 1];
    const current = throwsSignals[index];
    if (!previous || !current) {
      continue;
    }
    if (normalizeHand(previous.fromHand) !== normalizeHand(current.fromHand)) {
      handSwitches += 1;
    }
  }

  const alternationRatio =
    throwsSignals.length < 2 ? gaitProfile.lateralBias : handSwitches / (throwsSignals.length - 1);
  const lateralFit = clamp01(
    1 - Math.abs(alternationRatio - gaitProfile.lateralBias) / Math.max(0.05, gaitProfile.lateralBias)
  );

  return round(clamp01(0.7 * flightCloseness + 0.3 * lateralFit));
}

function deriveNoveltyScore(
  throwsSignals: readonly JugglingOrbitThrowSignal[],
  model: CompositionModel
): number {
  const candidateSet = new Set<string>(
    throwsSignals.map((throwSignal, index) =>
      templateSignature(toTemplate(throwSignal, `candidate-${index + 1}`))
    )
  );
  const candidateSize = Math.max(1, candidateSet.size);

  let maxSimilarity = 0;
  for (const seedPattern of model.seedPatterns) {
    let shared = 0;
    for (const signature of seedPattern.templateSignatures) {
      if (candidateSet.has(signature)) {
        shared += 1;
      }
    }
    const denominator = Math.max(candidateSize, seedPattern.templateSignatures.length);
    const similarity = denominator > 0 ? shared / denominator : 0;
    if (similarity > maxSimilarity) {
      maxSimilarity = similarity;
    }
  }

  return round(clamp01(1 - maxSimilarity));
}

function deriveMachineSuitabilityScore(
  orbit: JugglingOrbitFormalizationBundle
): number {
  const orbitCount = Math.max(1, orbit.aggregate.orbitCount);
  const machineOnlyRatio = orbit.aggregate.machineOnlyOrbitCount / orbitCount;
  const asymmetricRatio = orbit.aggregate.asymmetricMultiplexOrbitCount / orbitCount;
  const targetMachineRatio = 0.35;
  const ratioFit = clamp01(
    1 - Math.abs(machineOnlyRatio - targetMachineRatio) / (1 - targetMachineRatio)
  );

  return round(clamp01(0.7 * ratioFit + 0.3 * asymmetricRatio));
}

function deriveMpcProxyMetrics(
  throwsSignals: readonly JugglingOrbitThrowSignal[],
  orbit: JugglingOrbitFormalizationBundle,
  gaitProfile: NormalizedGaitProfile,
  mpcConfig: NormalizedMpcConfig
): JugglingMpcProxyMetrics {
  const throwCount = Math.max(1, throwsSignals.length);
  const maxFlight = throwsSignals.reduce(
    (max, throwSignal) => Math.max(max, normalizePositiveInt(throwSignal.flightBeats, 1)),
    1
  );
  const multiplexPressure = orbit.aggregate.maxMultiplexPerBeatHand / Math.max(1, throwCount / orbit.period);
  const horizonPressure = clamp01(
    Math.max(0, maxFlight - mpcConfig.horizonBeats) / Math.max(1, maxFlight)
  );
  const nonConvexRisk = clamp01(
    0.55 * (orbit.aggregate.asymmetricMultiplexOrbitCount / Math.max(1, orbit.aggregate.orbitCount)) +
      0.45 * clamp01(multiplexPressure / 1.5)
  );
  const dimensionalLoad = clamp01(
    (throwCount * (1 + orbit.aggregate.maxMultiplexPerBeatHand * 0.35)) / 36
  );
  const openRate = orbit.aggregate.openTransitionCount / Math.max(1, orbit.aggregate.transitionCount);
  const uncertaintyLoad = clamp01(
    gaitProfile.uncertaintySigma * 0.7 + openRate * 0.3
  );
  const budgetPressure = clamp01(
    (maxFlight * 2 + throwCount * 0.18) / Math.max(1, mpcConfig.solverBudgetMs * 1.3)
  );
  const restartGain = clamp01(mpcConfig.restartCount / 4);
  const warmStartGain = mpcConfig.warmStartEnabled
    ? clamp01(0.65 * restartGain + 0.35 * (1 - budgetPressure))
    : 0;

  return {
    horizonPressure: round(horizonPressure),
    nonConvexRisk: round(nonConvexRisk),
    dimensionalLoad: round(dimensionalLoad),
    uncertaintyLoad: round(uncertaintyLoad),
    warmStartGain: round(warmStartGain),
  };
}

function normalizeWeights(
  weights: Partial<JugglingCompositionWeights> | undefined
): JugglingCompositionWeights {
  const rawValidity = Math.max(
    0,
    normalizeFinite(weights?.validity, DEFAULT_WEIGHTS.validity)
  );
  const rawFlow = Math.max(0, normalizeFinite(weights?.flow, DEFAULT_WEIGHTS.flow));
  const rawNovelty = Math.max(
    0,
    normalizeFinite(weights?.novelty, DEFAULT_WEIGHTS.novelty)
  );
  const rawMachine = Math.max(
    0,
    normalizeFinite(
      weights?.machineSuitability,
      DEFAULT_WEIGHTS.machineSuitability
    )
  );

  const sum = rawValidity + rawFlow + rawNovelty + rawMachine;
  if (sum <= 0) {
    return DEFAULT_WEIGHTS;
  }

  return {
    validity: rawValidity / sum,
    flow: rawFlow / sum,
    novelty: rawNovelty / sum,
    machineSuitability: rawMachine / sum,
  };
}

function candidateSignature(throwsSignals: readonly JugglingOrbitThrowSignal[]): string {
  return throwsSignals
    .map((throwSignal, index) => {
      const template = toTemplate(throwSignal, `throw-${index + 1}`);
      return [
        `b${positiveModulo(throwSignal.beat, 4096)}`,
        `f${template.fromHand}`,
        `t${template.toHand ?? '-'}`,
        `h${template.flightBeats}`,
        `l${template.lane}`,
        `a${template.angleDeg.toFixed(1)}`,
        `s${template.speed.toFixed(2)}`,
      ].join(':');
    })
    .join('|');
}

function hashText(value: string): number {
  let hash = 2166136261;
  for (let index = 0; index < value.length; index += 1) {
    hash ^= value.charCodeAt(index);
    hash = Math.imul(hash, 16777619);
  }
  return hash >>> 0;
}

function deriveScore(
  throwsSignals: readonly JugglingOrbitThrowSignal[],
  orbit: JugglingOrbitFormalizationBundle,
  model: CompositionModel,
  weights: JugglingCompositionWeights,
  gaitProfile: NormalizedGaitProfile,
  mpcConfig: NormalizedMpcConfig
): JugglingCompositionScore {
  const throwCount = Math.max(1, orbit.aggregate.throwCount);
  const conservationMismatchRate = orbit.aggregate.conservationMismatchCount / throwCount;
  const openTransitionRate = orbit.aggregate.openTransitionCount / throwCount;
  const closedOrbitRatio =
    orbit.aggregate.orbitCount === 0
      ? 0
      : orbit.aggregate.closedOrbitCount / orbit.aggregate.orbitCount;

  const validity = round(
    clamp01(
      1 -
        conservationMismatchRate * 0.6 -
        openTransitionRate * 0.3 +
        closedOrbitRatio * 0.1
    )
  );
  const flow = deriveFlowScore(throwsSignals);
  const gaitCoupling = deriveGaitCouplingScore(throwsSignals, gaitProfile);
  const flowWithGait = round(clamp01(0.78 * flow + 0.22 * gaitCoupling));
  const novelty = deriveNoveltyScore(throwsSignals, model);
  const machineSuitability = deriveMachineSuitabilityScore(orbit);
  const mpc = deriveMpcProxyMetrics(throwsSignals, orbit, gaitProfile, mpcConfig);
  const mpcFeasibility = clamp01(
    1 -
      0.3 * mpc.horizonPressure -
      0.25 * mpc.nonConvexRisk -
      0.25 * mpc.dimensionalLoad -
      0.2 * mpc.uncertaintyLoad +
      0.25 * mpc.warmStartGain
  );
  const weightedBase = clamp01(
    weights.validity * validity +
      weights.flow * flowWithGait +
      weights.novelty * novelty +
      weights.machineSuitability * machineSuitability
  );
  const total = round(clamp01(0.9 * weightedBase + 0.1 * mpcFeasibility));

  return {
    validity,
    flow: flowWithGait,
    gaitCoupling,
    novelty,
    machineSuitability,
    mpc,
    total,
    penalties: {
      conservationMismatchRate: round(conservationMismatchRate),
      openTransitionRate: round(openTransitionRate),
    },
  };
}

function sortCandidates(
  left: JugglingCompositionCandidate,
  right: JugglingCompositionCandidate
): number {
  return (
    right.score.total - left.score.total ||
    right.score.validity - left.score.validity ||
    right.score.flow - left.score.flow ||
    right.score.novelty - left.score.novelty ||
    left.candidateId.localeCompare(right.candidateId)
  );
}

function evaluateGenome(
  genome: Genome,
  generation: number,
  index: number,
  period: number,
  model: CompositionModel,
  weights: JugglingCompositionWeights,
  gaitProfile: NormalizedGaitProfile,
  mpcConfig: NormalizedMpcConfig,
  source: string,
  generatedAt: string,
  input: JugglingChaosOrchestrationInput
): JugglingCompositionCandidate {
  const normalizedThrows = normalizeThrowSequence(genome.throws, period);
  const orbitInput = {
    period,
    throws: normalizedThrows,
    source,
    generatedAt,
    ...(input.machineOnlyMultiplexThreshold !== undefined
      ? {
          machineOnlyMultiplexThreshold: input.machineOnlyMultiplexThreshold,
        }
      : {}),
    ...(input.machineOnlyAngleThresholdDeg !== undefined
      ? {
          machineOnlyAngleThresholdDeg: input.machineOnlyAngleThresholdDeg,
        }
      : {}),
    ...(input.machineOnlySpeedRatioThreshold !== undefined
      ? {
          machineOnlySpeedRatioThreshold: input.machineOnlySpeedRatioThreshold,
        }
      : {}),
  };
  const orbit = deriveJugglingOrbitFormalization(orbitInput);
  const signature = candidateSignature(normalizedThrows);
  const hash = hashText(signature).toString(16).padStart(8, '0');
  const candidateId = `candidate:g${generation}:i${index}:h${hash}`;

  return {
    candidateId,
    signature,
    throws: normalizedThrows,
    orbit,
    score: deriveScore(
      normalizedThrows,
      orbit,
      model,
      weights,
      gaitProfile,
      mpcConfig
    ),
    lineage: {
      generation,
      parentCandidateIds: [...genome.parentCandidateIds],
      mutationEvents: genome.mutationEvents,
      seedPatternIds: [...new Set(genome.seedPatternIds)],
      proposerRoute: genome.proposerRoute,
    },
  };
}

function tournamentSelect(
  candidates: readonly JugglingCompositionCandidate[],
  tournamentSize: number,
  rng: () => number
): JugglingCompositionCandidate {
  const initial =
    pickOne(candidates, rng) ?? candidates[0];
  if (!initial) {
    throw new Error('tournamentSelect requires at least one candidate');
  }

  let best = initial;
  for (let iteration = 0; iteration < tournamentSize; iteration += 1) {
    const contender = pickOne(candidates, rng);
    if (!contender) {
      continue;
    }
    if (sortCandidates(contender, best) < 0) {
      best = contender;
    }
  }
  return best;
}

function mergePatternIds(
  left: readonly string[],
  right: readonly string[]
): readonly string[] {
  return [...new Set([...left, ...right])].sort((a, b) =>
    a.localeCompare(b)
  );
}

function buildGenerationSummary(
  generation: number,
  candidates: readonly JugglingCompositionCandidate[],
  previousBestScore: number | null
): JugglingCompositionGenerationSummary {
  const scores = candidates.map((candidate) => candidate.score.total);
  const best = candidates[0];
  const minScore = scores.reduce(
    (min, score) => (score < min ? score : min),
    scores[0] ?? 0
  );
  const maxScore = scores.reduce(
    (max, score) => (score > max ? score : max),
    scores[0] ?? 0
  );
  const diversity =
    candidates.length === 0
      ? 0
      : new Set(candidates.map((candidate) => candidate.signature)).size /
        candidates.length;
  const mutationEvents = candidates.map((candidate) => candidate.lineage.mutationEvents);

  let markovCount = 0;
  let remixCount = 0;
  let eliteCarryCount = 0;
  for (const candidate of candidates) {
    if (candidate.lineage.proposerRoute === 'markov') {
      markovCount += 1;
    } else if (candidate.lineage.proposerRoute === 'corpus-remix') {
      remixCount += 1;
    } else {
      eliteCarryCount += 1;
    }
  }

  const horizonPressureMean = mean(
    candidates.map((candidate) => candidate.score.mpc.horizonPressure)
  );
  const nonConvexRiskMean = mean(
    candidates.map((candidate) => candidate.score.mpc.nonConvexRisk)
  );
  const dimensionalLoadMean = mean(
    candidates.map((candidate) => candidate.score.mpc.dimensionalLoad)
  );
  const uncertaintyLoadMean = mean(
    candidates.map((candidate) => candidate.score.mpc.uncertaintyLoad)
  );

  return {
    generation,
    bestCandidateId: best?.candidateId ?? `candidate:g${generation}:none`,
    bestScore: best?.score.total ?? 0,
    convergenceSlope:
      previousBestScore === null
        ? 0
        : round((best?.score.total ?? 0) - previousBestScore),
    meanScore: round(mean(scores)),
    minScore: round(minScore),
    maxScore: round(maxScore),
    meanMutationEvents: round(mean(mutationEvents)),
    proposerMix: {
      markov: round(markovCount / Math.max(1, candidates.length)),
      corpusRemix: round(remixCount / Math.max(1, candidates.length)),
      eliteCarry: round(eliteCarryCount / Math.max(1, candidates.length)),
    },
    mpc: {
      horizonPressure: round(horizonPressureMean),
      nonConvexRisk: round(nonConvexRiskMean),
      dimensionalLoad: round(dimensionalLoadMean),
      uncertaintyLoad: round(uncertaintyLoadMean),
    },
    diversity: round(diversity),
  };
}

function topUniqueBySignature(
  candidates: readonly JugglingCompositionCandidate[],
  count: number
): readonly JugglingCompositionCandidate[] {
  const seen = new Set<string>();
  const output: JugglingCompositionCandidate[] = [];
  for (const candidate of [...candidates].sort(sortCandidates)) {
    if (seen.has(candidate.signature)) {
      continue;
    }
    seen.add(candidate.signature);
    output.push(candidate);
    if (output.length >= count) {
      break;
    }
  }
  return output;
}

/**
 * Handles the gnosis orchestrate Juggling Chaos Composition workflow.
 */
export function orchestrateJugglingChaosComposition(
  input: JugglingChaosOrchestrationInput
): JugglingChaosOrchestrationBundle {
  const period = normalizePositiveInt(input.period, 1);
  const seed = Math.floor(normalizeFinite(input.seed, DEFAULT_SEED));
  const populationSize = normalizePositiveInt(
    input.populationSize,
    DEFAULT_POPULATION_SIZE
  );
  const generations = normalizePositiveInt(input.generations, DEFAULT_GENERATIONS);
  const throwsPerPattern = normalizePositiveInt(
    input.throwsPerPattern,
    DEFAULT_THROWS_PER_PATTERN
  );
  const proposer = normalizeProposerId(input.proposer);
  const generatedAt = input.generatedAt ?? nowIso();
  const source = input.source ?? 'gnosis:juggling-chaos-orchestration';
  const weights = normalizeWeights(input.weights);
  const gaitProfile = normalizeGaitProfile(input.gaitProfile);
  const mpcConfig = normalizeMpcConfig(input.mpc);
  const rng = createRng(seed);
  const model = buildModel(input.corpus, period);
  const mutationProbability = mpcConfig.warmStartEnabled ? 0.52 : 0.67;

  const population: Genome[] = [];
  for (const seedPattern of model.seedPatterns) {
    if (population.length >= populationSize) {
      break;
    }
    population.push({
      throws: seedPattern.throws,
      parentCandidateIds: [],
      seedPatternIds: [seedPattern.patternId],
      mutationEvents: 0,
      proposerRoute: 'corpus-remix',
    });
  }

  while (population.length < populationSize) {
    const proposal = proposeInitialPattern(
      proposer,
      model,
      period,
      throwsPerPattern,
      rng
    );
    population.push({
      throws: proposal.throws,
      parentCandidateIds: [],
      seedPatternIds: [],
      mutationEvents: 0,
      proposerRoute: proposal.proposerRoute,
    });
  }

  const generationSummaries: JugglingCompositionGenerationSummary[] = [];
  const allCandidates: JugglingCompositionCandidate[] = [];
  let currentPopulation = population;
  let previousBestScore: number | null = null;

  for (let generation = 0; generation < generations; generation += 1) {
    const evaluated = currentPopulation
      .map((genome, index) =>
        evaluateGenome(
          genome,
          generation,
          index,
          period,
          model,
          weights,
          gaitProfile,
          mpcConfig,
          source,
          generatedAt,
          input
        )
      )
      .sort(sortCandidates);

    allCandidates.push(...evaluated);
    generationSummaries.push(
      buildGenerationSummary(generation, evaluated, previousBestScore)
    );
    previousBestScore = evaluated[0]?.score.total ?? previousBestScore;

    if (generation === generations - 1) {
      break;
    }

    const nextPopulation: Genome[] = [];
    const eliteCount = Math.max(1, Math.min(4, Math.floor(populationSize * 0.15)));

    for (let index = 0; index < eliteCount; index += 1) {
      const elite = evaluated[index];
      if (!elite) {
        continue;
      }
      nextPopulation.push({
        throws: cloneThrows(elite.throws),
        parentCandidateIds: [elite.candidateId],
        seedPatternIds: [...elite.lineage.seedPatternIds],
        mutationEvents: 0,
        proposerRoute: 'elite-carry',
      });
    }

    while (nextPopulation.length < populationSize) {
      const parentA = tournamentSelect(evaluated, 3, rng);
      const parentB = tournamentSelect(evaluated, 3, rng);
      const proposal = proposeOffspringPattern(
        proposer,
        parentA,
        parentB,
        model,
        period,
        throwsPerPattern,
        rng
      );
      let childThrows = proposal.throws;
      let mutationEvents = 0;

      if (rng() < mutationProbability) {
        const mutation = mutatePattern(childThrows, period, rng);
        childThrows = mutation.throws;
        mutationEvents = mutation.mutationEvents;
      }

      nextPopulation.push({
        throws: normalizeThrowSequence(childThrows, period),
        parentCandidateIds: [parentA.candidateId, parentB.candidateId],
        seedPatternIds: mergePatternIds(
          parentA.lineage.seedPatternIds,
          parentB.lineage.seedPatternIds
        ),
        mutationEvents,
        proposerRoute: proposal.proposerRoute,
      });
    }

    currentPopulation = nextPopulation;
  }

  const sortedAll = [...allCandidates].sort(sortCandidates);
  const winner = sortedAll[0];
  if (!winner) {
    throw new Error('Failed to produce a juggling composition candidate');
  }

  return {
    generatedAt,
    source,
    theoremIds: JUGGLING_CHAOS_ORCHESTRATION_THEOREM_IDS,
    dependentTheoremIds: JUGGLING_ORBIT_THEOREM_IDS,
    seed,
    period,
    configuration: {
      populationSize,
      generations,
      throwsPerPattern,
      proposer,
      gaitProfile,
      mpc: mpcConfig,
      weights,
    },
    corpusDigest: {
      patternCount: model.seedPatterns.length,
      uniqueThrowTemplateCount: model.uniqueTemplateCount,
      styleTagCount: model.styleTagCount,
    },
    generationSummaries,
    winner,
    finalists: topUniqueBySignature(sortedAll, 5),
  };
}
