// @a0n/gnosis/src/knotgraft-client.ts
/**
 * Knotgraft client SDK.
 *
 * Minimal typed wrapper around `POST /api/knotgraft` (the commercial
 * surface of the Prismatic Refraction wire protocol). Handles:
 *
 *   - bearer auth
 *   - octet-stream vs JSON encoding selection
 *   - automatic `Retry-After` back-off on 429
 *   - typed response parsing
 *   - optional streaming-chunk helper for large payloads
 *
 * Dependency-free. Works on Node ≥ 18, browsers, Deno, Bun, and
 * Cloudflare Workers — uses only `fetch`, `atob`/`btoa`, and typed
 * arrays.
 */

// ─── Types ──────────────────────────────────────────────────────────

/** One knotgraph-shaped entry (app-zone projection of a settled shard). */
export interface KnotgraftEntry {
  readonly regionId: number;
  readonly entityId: number;
  readonly payloadLength: number;
  readonly phase: number;
}

/** Portable SQL materialization, returned when `emitSql: true`. Params
 *  that were originally BLOB are already base64-encoded. */
export interface KnotgraftSqlBundle {
  readonly ddl: string;
  readonly indexes: ReadonlyArray<string>;
  readonly insertSql: string;
  readonly params: ReadonlyArray<ReadonlyArray<number | string>>;
}

/** Successful response body. */
export interface KnotgraftResponse {
  readonly blocks: number;
  readonly totalBytes: number;
  readonly lanesOccupied: number;
  readonly laneCount: number;
  readonly tenantHi: number;
  readonly tenantLo: number;
  readonly entries: ReadonlyArray<KnotgraftEntry>;
  readonly throughputMBs: number | null;
  readonly rateLimit: {
    readonly limitPerMinute: number;
    readonly remaining: number;
    readonly windowMs: number;
  };
  readonly sql?: KnotgraftSqlBundle;
}

export interface KnotgraftClientOptions {
  /** Base URL of the edgework-app deployment. Defaults to same-origin. */
  readonly baseUrl?: string;
  /** Bearer token identifying the caller. Required by the endpoint. */
  readonly token: string;
  /** `fetch` override — inject for tests / non-default environments. */
  readonly fetch?: typeof globalThis.fetch;
  /** Maximum automatic retry count on 429 responses. Default 3. */
  readonly maxRetries?: number;
  /** Cap on any single retry's wait, in ms. Default 30_000. */
  readonly maxRetryWaitMs?: number;
}

export interface IngestOptions {
  /** Lane count (1..16). Default 16 (knotgraph app-zone width). */
  readonly laneCount?: number;
  /** Emit portable SQL for DuckDB / D1 / sqlite-wasm. Default false. */
  readonly emitSql?: boolean;
  /** Force JSON transport (base64-wrapped). Default: octet-stream for
   *  `Uint8Array`, JSON for everything else. */
  readonly forceJson?: boolean;
}

export class KnotgraftError extends Error {
  readonly status: number;
  readonly body: unknown;
  constructor(message: string, status: number, body: unknown) {
    super(message);
    this.name = 'KnotgraftError';
    this.status = status;
    this.body = body;
  }
}

// ─── Base64 helpers (environment-portable) ──────────────────────────

function bytesToBase64(bytes: Uint8Array): string {
  let bin = '';
  for (let i = 0; i < bytes.length; i++) bin += String.fromCharCode(bytes[i]!);
  return btoa(bin);
}

// ─── Client ─────────────────────────────────────────────────────────

export class KnotgraftClient {
  private readonly baseUrl: string;
  private readonly token: string;
  private readonly fetchImpl: typeof globalThis.fetch;
  private readonly maxRetries: number;
  private readonly maxRetryWaitMs: number;

  constructor(options: KnotgraftClientOptions) {
    if (!options.token) {
      throw new Error('KnotgraftClient requires a bearer token');
    }
    this.baseUrl = options.baseUrl ?? '';
    this.token = options.token;
    this.fetchImpl = options.fetch ?? globalThis.fetch.bind(globalThis);
    this.maxRetries = options.maxRetries ?? 3;
    this.maxRetryWaitMs = options.maxRetryWaitMs ?? 30_000;
  }

  /** Ingest a byte payload through the knotgraft endpoint. Accepts
   *  `Uint8Array`, `ArrayBuffer`, or `number[]` (0-255). */
  async ingest(
    payload: Uint8Array | ArrayBuffer | ReadonlyArray<number>,
    options: IngestOptions = {},
  ): Promise<KnotgraftResponse> {
    const bytes = toBytes(payload);
    const laneCount = options.laneCount ?? 16;
    if (!Number.isInteger(laneCount) || laneCount < 1 || laneCount > 16) {
      throw new RangeError(`laneCount must be an integer in [1, 16], got ${laneCount}`);
    }
    const url = `${this.baseUrl}/api/knotgraft?k=${laneCount}${
      options.emitSql ? '&emitSql=1' : ''
    }`;

    const useJson = options.forceJson ?? false;
    const init: RequestInit = {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${this.token}`,
        ...(useJson
          ? { 'Content-Type': 'application/json' }
          : { 'Content-Type': 'application/octet-stream' }),
      },
      body: (useJson
        ? JSON.stringify({
            data: bytesToBase64(bytes),
            laneCount,
            emitSql: options.emitSql ?? false,
          })
        : bytes) as unknown as BodyInit,
    };

    return this.requestWithRetry(url, init, 0);
  }

  /** Server-side tier/limits preview. Useful to size `chunkSize` before
   *  streaming large payloads. */
  async describe(): Promise<{
    tier: string;
    limits: { requestsPerMinute: number; maxRequestBytes: number };
  }> {
    const res = await this.fetchImpl(`${this.baseUrl}/api/knotgraft`, {
      method: 'GET',
      headers: { Authorization: `Bearer ${this.token}` },
    });
    if (!res.ok) {
      throw new KnotgraftError(
        `GET /api/knotgraft failed: ${res.status}`,
        res.status,
        await res.text(),
      );
    }
    return (await res.json()) as {
      tier: string;
      limits: { requestsPerMinute: number; maxRequestBytes: number };
    };
  }

  /** Split a large payload into per-request chunks ≤ `chunkBytes` and
   *  ingest each sequentially. Returns one response per chunk. Does
   *  NOT reassemble across chunks — each chunk produces its own
   *  knotgraph blocks keyed on the same tenant. */
  async ingestChunked(
    payload: Uint8Array | ArrayBuffer | ReadonlyArray<number>,
    chunkBytes: number,
    options: IngestOptions = {},
  ): Promise<ReadonlyArray<KnotgraftResponse>> {
    const bytes = toBytes(payload);
    if (!Number.isInteger(chunkBytes) || chunkBytes <= 0) {
      throw new RangeError(`chunkBytes must be a positive integer, got ${chunkBytes}`);
    }
    const out: KnotgraftResponse[] = [];
    for (let off = 0; off < bytes.length; off += chunkBytes) {
      const slice = bytes.subarray(off, Math.min(off + chunkBytes, bytes.length));
      out.push(await this.ingest(slice, options));
    }
    return out;
  }

  private async requestWithRetry(
    url: string,
    init: RequestInit,
    attempt: number,
  ): Promise<KnotgraftResponse> {
    const res = await this.fetchImpl(url, init);

    if (res.status === 429) {
      if (attempt >= this.maxRetries) {
        throw new KnotgraftError(
          `Rate limited after ${attempt + 1} attempts`,
          429,
          await safeBody(res),
        );
      }
      const retryAfter = parseRetryAfter(res) ?? 1000;
      const waitMs = Math.min(retryAfter, this.maxRetryWaitMs);
      await sleep(waitMs);
      return this.requestWithRetry(url, init, attempt + 1);
    }

    if (!res.ok) {
      throw new KnotgraftError(
        `POST /api/knotgraft failed: ${res.status}`,
        res.status,
        await safeBody(res),
      );
    }

    return (await res.json()) as KnotgraftResponse;
  }
}

// ─── Helpers ────────────────────────────────────────────────────────

function toBytes(
  p: Uint8Array | ArrayBuffer | ReadonlyArray<number>,
): Uint8Array {
  if (p instanceof Uint8Array) return p;
  if (p instanceof ArrayBuffer) return new Uint8Array(p);
  // ReadonlyArray<number>
  const out = new Uint8Array(p.length);
  for (let i = 0; i < p.length; i++) out[i] = (p[i] ?? 0) & 0xff;
  return out;
}

function parseRetryAfter(res: Response): number | null {
  const header = res.headers.get('Retry-After');
  if (!header) return null;
  // `Retry-After` is in seconds per RFC 9110.
  const seconds = Number(header);
  if (!Number.isFinite(seconds) || seconds < 0) return null;
  return seconds * 1000;
}

async function safeBody(res: Response): Promise<unknown> {
  try {
    return await res.json();
  } catch {
    try {
      return await res.text();
    } catch {
      return null;
    }
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}
