// @a0n/gnosis/src/knotgraft-fabric-server.ts
/**
 * Knotgraft Fabric Server — Prismatic Refraction ingest over
 * HTTP + TCP Aeon Flow + UDP Aeon Flow.
 *
 * This module is now a thin adapter over `@a0n/aeon/fabric`. The
 * heavy lifting — three-transport server, self-describing
 * FlowFrame parsing (wall-compatible), JSON tunnel envelope with
 * raw-path auto-fallback, receiver-side content-address dedup,
 * optional Lacey-cipher wrapping — lives in aeon. This file
 * contributes:
 *
 *   1. `processIngest` — the Prismatic Refraction pipeline core.
 *      Pure function, unchanged. Used by the handler and by tests
 *      directly.
 *   2. `createKnotgraftHandler(config)` — returns an `AeonFabricHandler`
 *      that applies the auth gate, size cap, and ingest pipeline.
 *      Can be composed into any aeon fabric.
 *   3. `startKnotgraftFabric` — convenience wrapper that wires the
 *      handler through `startAeonFabric` with the Lacey cipher plumbed
 *      in from a passphrase. One call = three transports up.
 *
 * The old `handleHttpIngest(ctx)` pure HTTP handler is retained for
 * testing compatibility (same input/output as before) — it delegates
 * to the same core pipeline.
 *
 * Soundness contract: every ingest runs `wireToStorageIsSound`
 * in-process; on violation the response is 500 / VENT. This is the
 * runtime projection of `THM-WIRE-TO-STORAGE-IS-SOUND`.
 */

import {
  startAeonFabric,
  type AeonCipher,
  type AeonFabricHandle,
  type AeonFabricHandler,
} from '@a0n/aeon/fabric';
import {
  rotatingHelixEncryptFrame,
  rotatingHelixDecryptFrame,
  rotatingHelixKeyFromSeed,
  seedFromPassphrase,
  type RotatingHelixKey,
} from './bitwise-cipher';
import { createRawStream } from './prismatic-refraction';
import {
  ingestToKnotgraph,
  KNOTGRAPH_DIMS,
  type IngestContext,
  type KnotgraphShapedBlock,
  type TenantId,
} from './prismatic-refraction-knotgraph';
import { materializeIngest } from './prismatic-refraction-duckdb';

// ─── Config ──────────────────────────────────────────────────────────

export interface KnotgraftFabricConfig {
  readonly httpPort?: number;
  readonly flowTcpPort?: number;
  readonly flowUdpPort?: number;
  readonly defaultTenant: TenantId;
  readonly authToken?: string;
  readonly laneCount?: number;
  readonly emitSql?: boolean;
  readonly maxRequestBytes?: number;
  readonly maxUdpPayloadBytes?: number;
  readonly laceyCipher?: {
    readonly enabled?: boolean;
    readonly passphrase?: string;
  };
}

export const DEFAULT_HTTP_PORT = 7801;
export const DEFAULT_FLOW_TCP_PORT = 4244;
export const DEFAULT_FLOW_UDP_PORT = 4245;

// ─── Result types ────────────────────────────────────────────────────

export interface KnotgraftEntrySummary {
  readonly regionId: number;
  readonly entityId: number;
  readonly payloadLength: number;
  readonly phase: number;
}

export interface KnotgraftProcessResult {
  readonly blocks: number;
  readonly totalBytes: number;
  readonly lanesOccupied: number;
  readonly laneCount: number;
  readonly tenantHi: number;
  readonly tenantLo: number;
  readonly entries: ReadonlyArray<KnotgraftEntrySummary>;
  readonly sql?: {
    readonly ddl: string;
    readonly indexes: ReadonlyArray<string>;
    readonly insertSql: string;
    readonly params: ReadonlyArray<ReadonlyArray<number | string>>;
  };
}

// ─── Pipeline core (unchanged) ──────────────────────────────────────

/**
 * Handles the gnosis process Ingest workflow.
 */
export function processIngest(
  bytes: Uint8Array,
  tenant: TenantId,
  laneCount: number,
  emitSql: boolean,
): KnotgraftProcessResult {
  if (laneCount < 1 || laneCount > 16 || !Number.isInteger(laneCount)) {
    throw new RangeError(`laneCount must be an integer in [1, 16], got ${laneCount}`);
  }
  const stream = createRawStream(Array.from(bytes));
  const ctx: IngestContext = {
    tenant,
    typeTag: 0xdeadbeef,
    schemaKnotHash: 0xfacefeed,
    provenanceDid: 0,
    freshnessClock: Date.now() & 0x3fffffff,
  };
  const result = ingestToKnotgraph(stream, laneCount, ctx);

  if (!result.witness.lanesBounded || !result.witness.payloadPreserved) {
    throw new Error('soundness_violation: wireToStorageIsSound failed');
  }

  const entries: KnotgraftEntrySummary[] = result.blocks.map(
    (b: KnotgraphShapedBlock) => ({
      regionId: b.dims[KNOTGRAPH_DIMS.REGION_ID]!,
      entityId: b.dims[KNOTGRAPH_DIMS.ENTITY_ID]!,
      payloadLength: b.dims[KNOTGRAPH_DIMS.FIELD_0]!,
      phase: b.dims[KNOTGRAPH_DIMS.FIELD_1]!,
    }),
  );

  const out: KnotgraftProcessResult = {
    blocks: result.blocks.length,
    totalBytes: result.totalIngestedBytes,
    lanesOccupied: result.blocks.length,
    laneCount,
    tenantHi: tenant.hi,
    tenantLo: tenant.lo,
    entries,
  };

  if (emitSql) {
    const mat = materializeIngest(result);
    const params = mat.params.map((row) =>
      row.map((v) =>
        v instanceof Uint8Array ? bytesToBase64(v) : v,
      ),
    );
    return { ...out, sql: { ddl: mat.ddl, indexes: mat.indexes, insertSql: mat.insertSql, params } };
  }
  return out;
}

// Retained for backward-compat with tests — identical behavior, delegates
// to processIngest. Dedup is now handled upstream by the fabric.
export interface ProcessIngestMeta {
  readonly deduped: boolean;
}
/**
 * Handles the gnosis process Ingest With Meta workflow.
 */
export function processIngestWithMeta(
  bytes: Uint8Array,
  tenant: TenantId,
  laneCount: number,
  emitSql: boolean,
): { result: KnotgraftProcessResult; meta: ProcessIngestMeta } {
  const result = processIngest(bytes, tenant, laneCount, emitSql);
  return { result, meta: { deduped: false } };
}

function bytesToBase64(bytes: Uint8Array): string {
  let bin = '';
  for (let i = 0; i < bytes.length; i++) bin += String.fromCharCode(bytes[i]!);
  return btoa(bin);
}

// ─── Knotgraft-shaped handler (Aeon Fabric handler) ─────────────────

/** Create an `AeonFabricHandler` that applies the knotgraft policy
 *  (auth gate, size cap, tenant stamp) and runs the ingest pipeline. */
export function createKnotgraftHandler(
  config: KnotgraftFabricConfig,
): AeonFabricHandler {
  return async (request: Request): Promise<Response> => {
    const url = new URL(request.url);
    const path = url.pathname;
    const accepted = path === '/' || path === '/ingest' || path === '/api/knotgraft';

    if (request.method === 'GET' && accepted) {
      return Response.json({
        endpoint: 'knotgraft-fabric',
        transports: {
          http: config.httpPort ?? DEFAULT_HTTP_PORT,
          tcpFlow: config.flowTcpPort ?? DEFAULT_FLOW_TCP_PORT,
          udpFlow: config.flowUdpPort ?? DEFAULT_FLOW_UDP_PORT,
        },
        laneCount: config.laneCount ?? 16,
        maxRequestBytes: config.maxRequestBytes ?? 64 * 1024 * 1024,
        maxUdpPayloadBytes: config.maxUdpPayloadBytes ?? 1400,
        formal: {
          leanModule: 'ForkRaceFoldTheorems/PrismaticRefraction.lean',
          crownTheorem: 'THM-WIRE-TO-STORAGE-IS-SOUND',
        },
      });
    }
    if (request.method !== 'POST' || !accepted) {
      return Response.json({ error: 'not_found' }, { status: 404 });
    }

    if (config.authToken) {
      const auth = request.headers.get('authorization') ?? '';
      if (auth !== `Bearer ${config.authToken}`) {
        return Response.json({ error: 'unauthorized' }, { status: 401 });
      }
    }

    const body = new Uint8Array(await request.arrayBuffer());
    const maxRequestBytes = config.maxRequestBytes ?? 64 * 1024 * 1024;
    if (body.byteLength > maxRequestBytes) {
      return Response.json(
        { error: 'payload_too_large', maxRequestBytes },
        { status: 413 },
      );
    }

    const laneCount =
      Number(url.searchParams.get('k')) ||
      Number(request.headers.get('x-knotgraft-lane-count')) ||
      config.laneCount ||
      16;
    const emitSqlQuery =
      url.searchParams.get('emitSql') ?? url.searchParams.get('emit_sql') ?? '';
    const emitSql =
      emitSqlQuery === '1' ||
      emitSqlQuery.toLowerCase() === 'true' ||
      (request.headers.get('x-knotgraft-emit-sql') ?? '').toLowerCase() === 'true' ||
      config.emitSql === true;

    try {
      const result = processIngest(body, config.defaultTenant, laneCount, emitSql);
      return Response.json(result);
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err);
      if (msg.startsWith('soundness_violation')) {
        return Response.json(
          { error: 'internal_invariant_failed', message: msg },
          { status: 500 },
        );
      }
      return Response.json(
        { error: 'invalid_request', message: msg },
        { status: 400 },
      );
    }
  };
}

// ─── Pure handler context (for tests) ───────────────────────────────

export interface HttpHandlerContext {
  readonly config: KnotgraftFabricConfig;
  readonly method: string;
  readonly path: string;
  readonly headers: ReadonlyMap<string, string>;
  readonly body: Uint8Array;
}

export interface HttpHandlerResponse {
  readonly status: number;
  readonly headers: Record<string, string>;
  readonly body: Uint8Array;
}

/** Legacy pure HTTP handler. Kept for testing; delegates to the same
 *  core handler that the fabric uses. */
export async function handleHttpIngest(
  ctx: HttpHandlerContext,
): Promise<HttpHandlerResponse> {
  const headers: Record<string, string> = {};
  for (const [k, v] of ctx.headers) headers[k] = v;
  const url = new URL(ctx.path, 'http://x-ranger.internal');
  const request = new Request(url, {
    method: ctx.method,
    headers,
    body:
      ctx.method === 'GET' || ctx.method === 'HEAD'
        ? undefined
        : (ctx.body as unknown as BodyInit | undefined),
  });
  const handler = createKnotgraftHandler(ctx.config);
  const response = await handler(request);
  const outHeaders: Record<string, string> = {};
  response.headers.forEach((value: string, k: string) => (outHeaders[k] = value));
  const buf = new Uint8Array(await response.arrayBuffer());
  return { status: response.status, headers: outHeaders, body: buf };
}

// ─── Lacey cipher adapter ───────────────────────────────────────────

async function resolveCipher(
  config: KnotgraftFabricConfig,
): Promise<AeonCipher | undefined> {
  const enabled = config.laceyCipher?.enabled ?? true;
  if (!enabled) return undefined;
  const passphrase =
    config.laceyCipher?.passphrase ?? process.env['KNOTGRAFT_CIPHER_PASSPHRASE'] ?? null;
  if (!passphrase) return undefined;
  const seed = await seedFromPassphrase(passphrase);
  const key: RotatingHelixKey = rotatingHelixKeyFromSeed(seed);
  return {
    encryptFrame: (frame: Uint8Array) => rotatingHelixEncryptFrame(frame, key),
    decryptFrame: (frame: Uint8Array) => rotatingHelixDecryptFrame(frame, key),
  };
}

// ─── Server bootstrap ──────────────────────────────────────────────

export interface KnotgraftFabricHandle {
  readonly httpPort: number;
  readonly flowTcpPort: number | null;
  readonly flowUdpPort: number | null;
  close(): Promise<void>;
}

/**
 * Handles the gnosis start Knotgraft Fabric workflow.
 */
export async function startKnotgraftFabric(
  config: KnotgraftFabricConfig,
): Promise<KnotgraftFabricHandle> {
  const handler = createKnotgraftHandler(config);
  const cipher = await resolveCipher(config);

  const fabric: AeonFabricHandle = await startAeonFabric({
    handler,
    httpPort: config.httpPort ?? DEFAULT_HTTP_PORT,
    flowTcpPort: config.flowTcpPort ?? DEFAULT_FLOW_TCP_PORT,
    flowUdpPort: config.flowUdpPort ?? DEFAULT_FLOW_UDP_PORT,
    cipher,
    dedup: { enabled: true },
  });

  return {
    httpPort: fabric.httpPort ?? (config.httpPort ?? DEFAULT_HTTP_PORT),
    flowTcpPort: fabric.flowTcpPort,
    flowUdpPort: fabric.flowUdpPort,
    close: () => fabric.close(),
  };
}
