// @a0n/gnosis/src/knotgraft-diversity-client.ts
/**
 * Knotgraft Diversity Client — the Aeon model.
 *
 * Does NOT choose a transport. Fires the same payload over all
 * available transports simultaneously and takes the first arrival.
 * The receiver deduplicates by `(tenantHi, tenantLo, entityId)` —
 * content-addressed identity is deterministic (via
 * `race_deterministic`), so duplicate arrivals at the server are a
 * free confirmation, not wasted work.
 *
 * The client stagger-starts transports by measured/expected latency
 * (HTTP first because it's slowest) so all three arrive at the
 * receiver around the same moment. The first valid response back is
 * returned; late legs are not cancelled — they still deliver, and
 * the receiver's content-addressed dedup cache collapses them.
 *
 * This is the full expression of the Aeon diversity contract:
 *
 *   FORK  — client fires N legs in parallel
 *   RACE  — client takes the first valid response
 *   FOLD  — receiver unifies duplicates via content address
 *
 * No transport is canonical. The diverse dispatch is the contract.
 *
 * Node / Bun only (uses `node:net` + `node:dgram`).
 */

import { createConnection, type Socket } from 'node:net';
import { createSocket } from 'node:dgram';
import { FlowCodec } from '@a0n/aeon/flow';
import type { KnotgraftResponse } from './knotgraft-client';

// ─── Config ──────────────────────────────────────────────────────────

export interface KnotgraftDiversityConfig {
  readonly http: {
    readonly baseUrl: string;
    readonly token: string;
  };
  readonly tcpFlow?: {
    readonly host: string;
    readonly port: number;
  };
  readonly udpFlow?: {
    readonly host: string;
    readonly port: number;
  };
  /** HTTP head-start in ms. Because HTTP is the slowest leg, fire it
   *  first, then fire TCP and UDP after this delay so the three
   *  arrivals converge at the receiver. Default: 0 (fire all at once
   *  — simplest). */
  readonly httpHeadStartMs?: number;
  /** Per-leg timeout. Default 10_000 ms. */
  readonly legTimeoutMs?: number;
  /** Which transports to fire. Default: all three if configured. */
  readonly fire?: ReadonlyArray<'http' | 'tcpFlow' | 'udpFlow'>;
}

export interface DiversityLegReport {
  readonly transport: 'http' | 'tcpFlow' | 'udpFlow';
  readonly status: 'winner' | 'runner-up' | 'error' | 'timeout' | 'not-fired';
  readonly elapsedMs: number;
  readonly deduped?: boolean;
  readonly error?: string;
}

export interface DiversityIngestResult {
  readonly result: KnotgraftResponse;
  readonly winner: 'http' | 'tcpFlow' | 'udpFlow';
  readonly winnerElapsedMs: number;
  readonly legs: ReadonlyArray<DiversityLegReport>;
  readonly totalElapsedMs: number;
}

export interface IngestOptions {
  readonly laneCount?: number;
  readonly emitSql?: boolean;
}

// ─── Client ─────────────────────────────────────────────────────────

export class KnotgraftDiversityClient {
  private readonly config: KnotgraftDiversityConfig;
  private readonly codec: FlowCodec;

  constructor(config: KnotgraftDiversityConfig) {
    if (!config.http?.token) {
      throw new Error('KnotgraftDiversityClient requires http.token');
    }
    this.config = config;
    this.codec = FlowCodec.createSync();
  }

  async ingest(
    bytes: Uint8Array,
    options: IngestOptions = {},
  ): Promise<DiversityIngestResult> {
    const laneCount = options.laneCount ?? 16;
    const emitSql = options.emitSql ?? false;
    const fire = this.config.fire ?? this.availableTransports();
    const headStart = this.config.httpHeadStartMs ?? 0;
    const timeoutMs = this.config.legTimeoutMs ?? 10_000;

    const t0 = performance.now();
    const legs: Map<'http' | 'tcpFlow' | 'udpFlow', Promise<LegOutcome>> = new Map();

    if (fire.includes('http')) {
      legs.set(
        'http',
        this.fireHttp(bytes, laneCount, emitSql, timeoutMs, t0),
      );
    }
    if (fire.includes('tcpFlow') && this.config.tcpFlow) {
      legs.set(
        'tcpFlow',
        delay(headStart).then(() =>
          this.fireTcp(bytes, laneCount, emitSql, timeoutMs, t0),
        ),
      );
    }
    if (fire.includes('udpFlow') && this.config.udpFlow) {
      legs.set(
        'udpFlow',
        delay(headStart).then(() =>
          this.fireUdp(bytes, laneCount, emitSql, timeoutMs, t0),
        ),
      );
    }

    if (legs.size === 0) {
      throw new Error('no transports configured or fired');
    }

    // Race: first leg to resolve with a valid response wins.
    const winner = await firstValid(legs);

    // Collect all leg outcomes (including late arrivals) for reporting.
    const settled = await Promise.allSettled(
      [...legs.entries()].map(async ([transport, p]) => ({
        transport,
        outcome: await p,
      })),
    );

    const legsReport: DiversityLegReport[] = settled.map((s) => {
      if (s.status === 'fulfilled') {
        const { transport, outcome } = s.value;
        return {
          transport,
          status:
            transport === winner.transport
              ? 'winner'
              : outcome.ok
                ? 'runner-up'
                : outcome.timedOut
                  ? 'timeout'
                  : 'error',
          elapsedMs: outcome.elapsedMs,
          deduped: outcome.ok ? outcome.deduped : undefined,
          error: outcome.error,
        };
      }
      return {
        transport: 'http',
        status: 'error',
        elapsedMs: 0,
        error: 'promise rejected',
      };
    });

    return {
      result: winner.result,
      winner: winner.transport,
      winnerElapsedMs: winner.elapsedMs,
      legs: legsReport,
      totalElapsedMs: performance.now() - t0,
    };
  }

  private availableTransports(): ReadonlyArray<'http' | 'tcpFlow' | 'udpFlow'> {
    const out: Array<'http' | 'tcpFlow' | 'udpFlow'> = ['http'];
    if (this.config.tcpFlow) out.push('tcpFlow');
    if (this.config.udpFlow) out.push('udpFlow');
    return out;
  }

  private async fireHttp(
    bytes: Uint8Array,
    laneCount: number,
    emitSql: boolean,
    timeoutMs: number,
    t0: number,
  ): Promise<LegOutcome> {
    const url = `${this.config.http.baseUrl}/api/knotgraft?k=${laneCount}${emitSql ? '&emitSql=1' : ''}`;
    const controller = new AbortController();
    const tid = setTimeout(() => controller.abort(), timeoutMs);
    try {
      const res = await fetch(url, {
        method: 'POST',
        headers: {
          Authorization: `Bearer ${this.config.http.token}`,
          'Content-Type': 'application/octet-stream',
        },
        body: bytes as BodyInit,
        signal: controller.signal,
      });
      clearTimeout(tid);
      if (!res.ok) {
        return {
          ok: false,
          timedOut: false,
          elapsedMs: performance.now() - t0,
          error: `HTTP ${res.status}`,
        };
      }
      const result = (await res.json()) as KnotgraftResponse;
      const deduped = res.headers.get('x-knotgraft-deduped') === 'true';
      return {
        ok: true,
        timedOut: false,
        result,
        deduped,
        elapsedMs: performance.now() - t0,
      };
    } catch (err) {
      clearTimeout(tid);
      const timedOut = (err as Error)?.name === 'AbortError';
      return {
        ok: false,
        timedOut,
        elapsedMs: performance.now() - t0,
        error: (err as Error)?.message ?? 'fetch failed',
      };
    }
  }

  private fireTcp(
    bytes: Uint8Array,
    _laneCount: number,
    _emitSql: boolean,
    timeoutMs: number,
    t0: number,
  ): Promise<LegOutcome> {
    const cfg = this.config.tcpFlow!;
    const codec = this.codec;
    return new Promise((resolve) => {
      const socket: Socket = createConnection(cfg.port, cfg.host);
      const tid = setTimeout(() => {
        socket.destroy();
        resolve({
          ok: false,
          timedOut: true,
          elapsedMs: performance.now() - t0,
          error: 'tcp timeout',
        });
      }, timeoutMs);
      let rx: Buffer = Buffer.alloc(0);

      socket.once('connect', () => {
        const frame = codec.encode({
          streamId: 1,
          sequence: 0,
          flags: 0,
          payload: bytes,
        });
        const lenBuf = Buffer.alloc(4);
        lenBuf.writeUInt32BE(frame.length);
        socket.write(lenBuf);
        socket.write(frame);
      });

      socket.on('data', (chunk: Buffer) => {
        rx = Buffer.concat([rx, chunk]);
        while (rx.length >= 4) {
          const msgLen = rx.readUInt32BE(0);
          if (rx.length < 4 + msgLen) break;
          const frameBytes = new Uint8Array(rx.subarray(4, 4 + msgLen));
          rx = rx.subarray(4 + msgLen);
          const decoded = codec.decode(frameBytes);
          if (decoded.frame.flags === 0 && decoded.frame.payload.length > 0) {
            clearTimeout(tid);
            socket.end();
            const result = JSON.parse(
              new TextDecoder().decode(decoded.frame.payload),
            ) as KnotgraftResponse;
            resolve({
              ok: true,
              timedOut: false,
              result,
              deduped: false,
              elapsedMs: performance.now() - t0,
            });
            return;
          }
        }
      });

      socket.on('error', (err) => {
        clearTimeout(tid);
        resolve({
          ok: false,
          timedOut: false,
          elapsedMs: performance.now() - t0,
          error: err.message,
        });
      });
    });
  }

  private fireUdp(
    bytes: Uint8Array,
    _laneCount: number,
    _emitSql: boolean,
    timeoutMs: number,
    t0: number,
  ): Promise<LegOutcome> {
    const cfg = this.config.udpFlow!;
    const codec = this.codec;
    return new Promise((resolve) => {
      const socket = createSocket('udp4');
      const tid = setTimeout(() => {
        socket.close();
        resolve({
          ok: false,
          timedOut: true,
          elapsedMs: performance.now() - t0,
          error: 'udp timeout',
        });
      }, timeoutMs);

      socket.once('message', (msg) => {
        clearTimeout(tid);
        socket.close();
        try {
          const decoded = codec.decode(new Uint8Array(msg));
          const result = JSON.parse(
            new TextDecoder().decode(decoded.frame.payload),
          ) as KnotgraftResponse;
          resolve({
            ok: true,
            timedOut: false,
            result,
            deduped: false,
            elapsedMs: performance.now() - t0,
          });
        } catch (err) {
          resolve({
            ok: false,
            timedOut: false,
            elapsedMs: performance.now() - t0,
            error: (err as Error).message,
          });
        }
      });

      socket.on('error', (err) => {
        clearTimeout(tid);
        resolve({
          ok: false,
          timedOut: false,
          elapsedMs: performance.now() - t0,
          error: err.message,
        });
      });

      const frame = codec.encode({
        streamId: 1,
        sequence: 0,
        flags: 0,
        payload: bytes,
      });
      socket.send(frame, cfg.port, cfg.host, (err) => {
        if (err) {
          clearTimeout(tid);
          resolve({
            ok: false,
            timedOut: false,
            elapsedMs: performance.now() - t0,
            error: err.message,
          });
        }
      });
    });
  }
}

// ─── Internal types ─────────────────────────────────────────────────

type LegOutcome =
  | {
      ok: true;
      timedOut: false;
      result: KnotgraftResponse;
      deduped: boolean;
      elapsedMs: number;
      error?: undefined;
    }
  | {
      ok: false;
      timedOut: boolean;
      elapsedMs: number;
      error: string;
      result?: undefined;
      deduped?: undefined;
    };

function delay(ms: number): Promise<void> {
  if (ms <= 0) return Promise.resolve();
  return new Promise((r) => setTimeout(r, ms));
}

async function firstValid(
  legs: Map<'http' | 'tcpFlow' | 'udpFlow', Promise<LegOutcome>>,
): Promise<{
  transport: 'http' | 'tcpFlow' | 'udpFlow';
  result: KnotgraftResponse;
  elapsedMs: number;
}> {
  return new Promise((resolve, reject) => {
    let remaining = legs.size;
    for (const [transport, p] of legs) {
      p.then((outcome) => {
        if (outcome.ok) {
          resolve({ transport, result: outcome.result, elapsedMs: outcome.elapsedMs });
        } else {
          remaining -= 1;
          if (remaining === 0) {
            reject(new Error('all transport legs failed'));
          }
        }
      }).catch(() => {
        remaining -= 1;
        if (remaining === 0) {
          reject(new Error('all transport legs failed'));
        }
      });
    }
  });
}
