// @a0n/gnosis/src/law.ts
import {
  attachUniverseRuntimeMetadata,
  createLawState,
  evaluateLawProgram,
  sha256Hex,
  type LawCapabilitySummary,
  type LawProgram,
  type LawState,
  type UniverseAdmissionReport,
} from './universe/index';
import type { GraphAST } from './betty/compiler';
import type { CapabilityContractSummary } from './capabilities/summary';

interface GgProgramNodeLike {
  id: string;
  labels: readonly string[];
  properties: Readonly<Record<string, string>>;
}

interface GgProgramEdgeLike {
  sourceIds: readonly string[];
  targetIds: readonly string[];
  type: string;
  properties: Readonly<Record<string, string>>;
}

interface GgProgramLike {
  nodes: readonly GgProgramNodeLike[];
  edges: readonly GgProgramEdgeLike[];
}

export interface GnosisLawProgramOptions {
  readonly subject?: string;
  readonly source?: string;
  readonly sourceDigest?: string;
  readonly capabilitySummary?: LawCapabilitySummary;
  readonly dependencyCertificateDigests?: readonly string[];
  readonly metadata?: Readonly<Record<string, string | number | boolean>>;
}

export interface GnosisUniverseAdmissionOptions {
  readonly correctnessOk: boolean;
  readonly capabilityOk: boolean;
  readonly reasons?: readonly string[];
}

export type GnosisUniverseAdmissionReport = UniverseAdmissionReport;

function summarizeCapabilities(
  capabilitySummary?: CapabilityContractSummary | LawCapabilitySummary,
  target?: string,
  issueCount?: number
): LawCapabilitySummary {
  if (!capabilitySummary) {
    return target || issueCount !== undefined
      ? {
          ...(target !== undefined && { target }),
          ...(issueCount !== undefined && { issueCount }),
        }
      : {};
  }

  if (
    'required' in capabilitySummary &&
    Array.isArray(capabilitySummary.required)
  ) {
    return {
      ...(target !== undefined && { target }),
      required: [...capabilitySummary.required],
      ...(Array.isArray(capabilitySummary.declared)
        ? { declared: [...capabilitySummary.declared] }
        : {}),
      ...(Array.isArray(capabilitySummary.inferred)
        ? { inferred: [...capabilitySummary.inferred] }
        : {}),
      ...(issueCount !== undefined ? { issueCount } : {}),
      ...('importedSurface' in capabilitySummary &&
      Array.isArray(capabilitySummary.importedSurface)
        ? { importedSurface: [...capabilitySummary.importedSurface] }
        : {}),
    };
  }

  return {
    ...(target !== undefined && { target }),
    ...(issueCount !== undefined ? { issueCount } : {}),
  };
}

function toSourceDigest(options: GnosisLawProgramOptions): string | undefined {
  if (options.sourceDigest) {
    return options.sourceDigest;
  }
  if (typeof options.source === 'string') {
    return sha256Hex(options.source);
  }
  return undefined;
}

/**
 * Handles the gnosis graph Ast To Law Program workflow.
 */
export function graphAstToLawProgram(
  ast: GraphAST,
  options: GnosisLawProgramOptions = {}
): LawProgram {
  return {
    nodes: new Map(
      [...ast.nodes.entries()].map(([nodeId, node]) => [
        nodeId,
        {
          id: node.id,
          labels: [...node.labels],
          properties: { ...node.properties },
        },
      ])
    ),
    edges: ast.edges.map((edge) => ({
      sourceIds: [...edge.sourceIds],
      targetIds: [...edge.targetIds],
      type: edge.type,
      properties: { ...edge.properties },
    })),
    ...(options.subject ? { subject: options.subject } : {}),
    ...(toSourceDigest(options)
      ? { sourceDigest: toSourceDigest(options)! }
      : {}),
    ...(options.capabilitySummary
      ? { capabilitySummary: options.capabilitySummary }
      : {}),
    ...(options.dependencyCertificateDigests
      ? {
          dependencyCertificateDigests: [
            ...options.dependencyCertificateDigests,
          ],
        }
      : {}),
    ...(options.metadata ? { metadata: options.metadata } : {}),
  };
}

/**
 * Handles the gnosis gg Program To Law Program workflow.
 */
export function ggProgramToLawProgram(
  program: GgProgramLike,
  options: GnosisLawProgramOptions = {}
): LawProgram {
  return {
    nodes: new Map(
      program.nodes.map((node) => [
        node.id,
        {
          id: node.id,
          labels: [...node.labels],
          properties: { ...node.properties },
        },
      ])
    ),
    edges: program.edges.map((edge) => ({
      sourceIds: [...edge.sourceIds],
      targetIds: [...edge.targetIds],
      type: edge.type,
      properties: { ...edge.properties },
    })),
    ...(options.subject ? { subject: options.subject } : {}),
    ...(toSourceDigest(options)
      ? { sourceDigest: toSourceDigest(options)! }
      : {}),
    ...(options.capabilitySummary
      ? { capabilitySummary: options.capabilitySummary }
      : {}),
    ...(options.dependencyCertificateDigests
      ? {
          dependencyCertificateDigests: [
            ...options.dependencyCertificateDigests,
          ],
        }
      : {}),
    ...(options.metadata ? { metadata: options.metadata } : {}),
  };
}

/**
 * Handles the gnosis certify Gnosis Ast workflow.
 */
export function certifyGnosisAst(
  ast: GraphAST,
  options: GnosisLawProgramOptions = {}
): ReturnType<typeof evaluateLawProgram> {
  return evaluateLawProgram(graphAstToLawProgram(ast, options));
}

/**
 * Handles the gnosis certify Gnosis Program workflow.
 */
export function certifyGnosisProgram(
  program: GgProgramLike,
  options: GnosisLawProgramOptions = {}
): ReturnType<typeof evaluateLawProgram> {
  return evaluateLawProgram(ggProgramToLawProgram(program, options));
}

/**
 * Creates the Universe Admission Report.
 */
export function createUniverseAdmissionReport(
  evaluation: ReturnType<typeof evaluateLawProgram>,
  options: GnosisUniverseAdmissionOptions
): UniverseAdmissionReport {
  const reasons = [...(options.reasons ?? [])];
  if (!options.correctnessOk) {
    reasons.push('Formal correctness is not verified.');
  }
  if (!options.capabilityOk) {
    reasons.push(
      'Capability contract is not admissible for the target runtime.'
    );
  }
  if (evaluation.verdict.status !== 'verified') {
    reasons.push(`Universe verdict is ${evaluation.verdict.status}.`);
  }

  return {
    kernelId: evaluation.certificate.kernelId,
    status: evaluation.verdict.status,
    ok:
      options.correctnessOk &&
      options.capabilityOk &&
      evaluation.verdict.status === 'verified',
    correctnessOk: options.correctnessOk,
    capabilityOk: options.capabilityOk,
    reasons,
    certificate: evaluation.certificate,
    law: evaluation.law,
  };
}

/**
 * Creates the Law State From Certificate.
 */
export function createLawStateFromCertificate(
  certificate: ReturnType<typeof evaluateLawProgram>['certificate']
): LawState {
  return createLawState(certificate);
}

/**
 * Handles the gnosis attach Runtime Metadata To Certificate workflow.
 */
export function attachRuntimeMetadataToCertificate(
  certificate: ReturnType<typeof evaluateLawProgram>['certificate'],
  runtimeMetadata: {
    executedAt: number;
    entrypoint?: string;
    runtimeName?: string;
  }
): ReturnType<typeof evaluateLawProgram>['certificate'] {
  return attachUniverseRuntimeMetadata(certificate, runtimeMetadata);
}

/**
 * Converts input into Law Capability Summary.
 */
export function toLawCapabilitySummary(
  capabilitySummary?: CapabilityContractSummary | LawCapabilitySummary,
  target?: string,
  issueCount?: number
): LawCapabilitySummary {
  return summarizeCapabilities(capabilitySummary, target, issueCount);
}

/**
 * Formats the Universe Admission Reasons.
 */
export function formatUniverseAdmissionReasons(
  report: UniverseAdmissionReport
): string[] {
  return report.reasons.length > 0
    ? [...report.reasons]
    : [`Universe verdict is ${report.status}.`];
}

/**
 * Handles the gnosis assert Universe Admission workflow.
 */
export function assertUniverseAdmission(
  report: UniverseAdmissionReport,
  context: string
): void {
  if (report.ok || process.env['A0_FORCE'] === 'true') {
    return;
  }

  throw new Error(
    `${context} rejected by universe admission: ${formatUniverseAdmissionReasons(
      report
    ).join(' ')}\nChecks: ${report.certificate.clockwork.checks
      .map(c => `${c.id}: ${c.pass ? 'PASS' : 'FAIL'} (${c.theorem ?? 'No details'})`)
      .join('\n')}`
  );
}
