// @a0n/gnosis/src/lean-execution.ts
import {
  GnosisWasmExecutionSurface,
  type ExecuteWasmArtifactRequest,
  type GnosisWasmExecutionSurfaceOptions,
  type RegisterWasmArtifactRequest,
  type WasmExecutionControlRequest,
  type WasmExecutionControlResponse,
  type WasmExecutionResponse,
  type WasmExecutionSurfaceStatus,
} from './wasm-execution';

export const LEAN_EXECUTION_CONTRACT = 'lean-wasm-json-v1';

export class LeanExecutorNotRegisteredError extends Error {
  readonly code = 'LEAN_EXECUTOR_NOT_REGISTERED';

  constructor() {
    super(
      'Lean execution requires a registered Lean WASM executor. Register one through /.aeon/lean/register before executing.'
    );
    this.name = 'LeanExecutorNotRegisteredError';
  }
}

export interface GnosisLeanExecutionSurfaceOptions {
  readonly enabled?: boolean;
  readonly basePath?: string;
  readonly wasmSurface?: GnosisWasmExecutionSurface | null;
  readonly wasmExecution?: GnosisWasmExecutionSurfaceOptions;
}

export interface RegisterLeanExecutorRequest
  extends RegisterWasmArtifactRequest {
  readonly contract?: typeof LEAN_EXECUTION_CONTRACT;
}

export interface ExecuteLeanBundleRequest {
  readonly bundle: unknown;
  readonly executionAuth?: ExecuteWasmArtifactRequest['executionAuth'];
  readonly executionAuthSource?: ExecuteWasmArtifactRequest['executionAuthSource'];
}

export interface LeanExecutionResponse {
  readonly contract: typeof LEAN_EXECUTION_CONTRACT;
  readonly executorArtifactId: string;
  readonly payload: unknown;
  readonly wasm: WasmExecutionResponse;
}

export interface LeanExecutionSurfaceStatus {
  readonly enabled: boolean;
  readonly contract: typeof LEAN_EXECUTION_CONTRACT;
  readonly executorRegistered: boolean;
  readonly executorArtifactId?: string;
  readonly reason?: 'no-lean-wasm-executor-registered';
  readonly wasm: WasmExecutionSurfaceStatus;
}

function stringifyJson(value: unknown): string {
  return JSON.stringify(
    value,
    (_key, candidate) => {
      if (typeof candidate === 'bigint') {
        return candidate.toString();
      }

      if (candidate instanceof Uint8Array) {
        return Array.from(candidate);
      }

      return candidate;
    },
    2
  );
}

function jsonResponse(
  status: number,
  body: unknown
): WasmExecutionControlResponse {
  return {
    status,
    headers: {
      'Content-Type': 'application/json',
      'Cache-Control': 'no-store',
    },
    body: new TextEncoder().encode(stringifyJson(body)),
  };
}

function normalizeRequestPath(request: WasmExecutionControlRequest): string {
  const path = request.path ?? '/';
  const queryIndex = path.indexOf('?');
  return queryIndex >= 0 ? path.slice(0, queryIndex) || '/' : path;
}

function parseJsonBody<T>(request: WasmExecutionControlRequest): T {
  if (!request.body || request.body.length === 0) {
    throw new Error('Request body is required.');
  }

  return JSON.parse(
    new TextDecoder('utf-8', { fatal: true }).decode(request.body)
  ) as T;
}

function normalizeBasePath(basePath: string | undefined): string {
  const raw = basePath?.trim() || '/.aeon/lean';
  const withLeadingSlash = raw.startsWith('/') ? raw : `/${raw}`;
  return withLeadingSlash.replace(/\/+$/, '') || '/.aeon/lean';
}

export class GnosisLeanExecutionSurface {
  private readonly enabled: boolean;
  private readonly basePath: string;
  private readonly wasmSurface: GnosisWasmExecutionSurface;
  private executorArtifactId: string | null = null;

  constructor(options: GnosisLeanExecutionSurfaceOptions = {}) {
    this.enabled = options.enabled ?? false;
    this.basePath = normalizeBasePath(options.basePath);
    this.wasmSurface =
      options.wasmSurface ??
      new GnosisWasmExecutionSurface({
        enabled: true,
        ...(options.wasmExecution ?? {}),
      });
  }

  registerExecutor(request: RegisterLeanExecutorRequest): string {
    if (
      request.contract !== undefined &&
      request.contract !== LEAN_EXECUTION_CONTRACT
    ) {
      throw new Error(
        `Unsupported Lean execution contract '${request.contract}'. Expected '${LEAN_EXECUTION_CONTRACT}'.`
      );
    }

    const artifact = this.wasmSurface.register({
      ...request,
      abi: request.abi ?? 'json-v1',
    });
    this.executorArtifactId = artifact.artifactId;
    return artifact.artifactId;
  }

  async execute(
    request: ExecuteLeanBundleRequest
  ): Promise<LeanExecutionResponse> {
    if (!this.executorArtifactId) {
      throw new LeanExecutorNotRegisteredError();
    }

    const wasm = await this.wasmSurface.execute({
      artifactId: this.executorArtifactId,
      input: request.bundle,
      ...(request.executionAuth
        ? { executionAuth: request.executionAuth }
        : {}),
      ...(request.executionAuthSource
        ? { executionAuthSource: request.executionAuthSource }
        : {}),
    });

    return {
      contract: LEAN_EXECUTION_CONTRACT,
      executorArtifactId: this.executorArtifactId,
      payload: wasm.payload,
      wasm,
    };
  }

  getStatus(): LeanExecutionSurfaceStatus {
    return {
      enabled: this.enabled,
      contract: LEAN_EXECUTION_CONTRACT,
      executorRegistered: this.executorArtifactId !== null,
      ...(this.executorArtifactId
        ? { executorArtifactId: this.executorArtifactId }
        : { reason: 'no-lean-wasm-executor-registered' as const }),
      wasm: this.wasmSurface.getStatus(),
    };
  }

  async handleRequest(
    request: WasmExecutionControlRequest
  ): Promise<WasmExecutionControlResponse | null> {
    if (!this.enabled) {
      return null;
    }

    const pathname = normalizeRequestPath(request);

    if (pathname === `${this.basePath}/status` && request.method === 'GET') {
      return jsonResponse(200, this.getStatus());
    }

    if (pathname === `${this.basePath}/register` && request.method === 'POST') {
      try {
        const body = parseJsonBody<RegisterLeanExecutorRequest>(request);
        const executorArtifactId = this.registerExecutor(body);
        return jsonResponse(200, {
          contract: LEAN_EXECUTION_CONTRACT,
          executorArtifactId,
        });
      } catch (error) {
        return jsonResponse(400, {
          error: error instanceof Error ? error.message : String(error),
        });
      }
    }

    if (pathname === `${this.basePath}/execute` && request.method === 'POST') {
      try {
        const body = parseJsonBody<ExecuteLeanBundleRequest>(request);
        const result = await this.execute(body);
        return jsonResponse(200, result);
      } catch (error) {
        if (error instanceof LeanExecutorNotRegisteredError) {
          return jsonResponse(409, {
            code: error.code,
            error: error.message,
          });
        }

        return jsonResponse(400, {
          error: error instanceof Error ? error.message : String(error),
        });
      }
    }

    return null;
  }
}
