// @a0n/gnosis/src/liquid-memory-topology.ts
/**
 * Liquid Memory Topology: Runtime Thermodynamic Memory Management
 *
 * Implements the formal Lean model `LiquidMemoryTopology.lean` in TypeScript.
 * Instead of static memory caps, enforces thermodynamically-bounded memory:
 * - held + vent = capacity + 2 (God Formula conservation law)
 * - information either flows (routes to neighbors) or vents (erased)
 * - pressure gradients drive mesh routing
 *
 * Based on: Gnosis.LiquidMemoryTopology (Lean 4)
 */

/**
 * Budget state for a single memory node.
 * Satisfies God Formula: memHeld(capacity, ventRate) + memVent(ventRate) = capacity + 2
 */
export interface MemoryBudget {
  /** Total capacity (R) */
  capacity: number;
  /** Currently held memory (w) */
  held: number;
  /** Active vent rate (v) */
  ventRate: number;
}

/**
 * A single item stored in liquid memory.
 */
export interface LiquidItem<T> {
  /** Unique key for this item */
  key: string;
  /** The stored value */
  value: T;
  /** Insertion timestamp (ms) */
  insertedAt: number;
  /** Information density (0.0 = low-value, 1.0 = high-value) */
  informationDensity: number;
  /** Accumulated storage cost (Landauer heat) */
  heatDebt: number;
}

/**
 * A node in the distributed liquid memory mesh.
 * Each node has a finite budget and routes/vents memory based on pressure.
 */
export interface LiquidMemoryNode<T> {
  /** Node identifier */
  id: string;
  /** Memory budget (capacity, held, ventRate) */
  budget: MemoryBudget;
  /** Items currently stored at this node */
  items: Map<string, LiquidItem<T>>;
  /** Callback to get neighboring nodes (lower pressure preferred) */
  neighbors: () => LiquidMemoryNode<T>[];
}

/**
 * Decision output for handling a memory item: hold, vent, or route.
 */
export type MemoryFlowDecision =
  | { action: 'hold' }
  | { action: 'vent'; heatReleased: number }
  | { action: 'route'; toNode: LiquidMemoryNode<unknown> };

// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
// Core God Formula Implementation
// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

/**
 * Memory held at a node.
 * God Formula: w(R, v) = R - min(v, R) + 1
 *
 * As vent rate increases, held decreases (saturation model).
 */
export function memHeld(capacity: number, ventRate: number): number {
  return capacity - Math.min(ventRate, capacity) + 1;
}

/**
 * Effective vent rate.
 * v_eff(v) = v + 1 (baseline heat dissipation)
 */
export function memVent(ventRate: number): number {
  return ventRate + 1;
}

/**
 * Verify conservation law: w + v = R + 2
 * (For testing / validation only.)
 */
export function verifyConservation(
  capacity: number,
  ventRate: number
): boolean {
  const w = memHeld(capacity, ventRate);
  const v = memVent(ventRate);
  return w + v === capacity + 2;
}

// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
// Pressure and Routing
// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

/**
 * Memory pressure: how close to capacity (0 = empty, 1.0 = full).
 *
 * Normalized to [0, 100] for decidability.
 */
export function computeMemoryPressure(node: LiquidMemoryNode<unknown>): number {
  if (node.budget.capacity === 0) return 0;
  return (node.budget.held * 100) / node.budget.capacity;
}

/**
 * Pressure gradient between two nodes.
 * Positive = node_a should route to node_b (pressure gradient favors it).
 */
export function pressureGradient(
  a: LiquidMemoryNode<unknown>,
  b: LiquidMemoryNode<unknown>
): number {
  return computeMemoryPressure(a) - computeMemoryPressure(b);
}

/**
 * Find the lowest-pressure neighbor (ideal route target).
 */
export function findLowPressureNeighbor<T>(
  node: LiquidMemoryNode<T>
): LiquidMemoryNode<T> | null {
  const neighbors = node.neighbors();
  if (neighbors.length === 0) return null;

  let best = neighbors[0];
  let bestPressure = computeMemoryPressure(best);

  for (let i = 1; i < neighbors.length; i++) {
    const pressure = computeMemoryPressure(neighbors[i]);
    if (pressure < bestPressure) {
      best = neighbors[i];
      bestPressure = pressure;
    }
  }

  return best;
}

// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
// Liquid Flow Decisions
// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

/**
 * Decide whether to hold, vent, or route a memory item.
 *
 * Logic:
 * 1. If node pressure < 50%, hold (plenty of capacity)
 * 2. If node pressure 50-80% and high density, hold
 * 3. If node pressure > 80%, must choose vent or route
 *    - If item density is low, vent (free up space via heat)
 *    - Else route to lower-pressure neighbor (if available)
 *    - If no neighbor available, vent as fallback
 */
export function decideLiquidFlow<T>(
  node: LiquidMemoryNode<T>,
  item: LiquidItem<T>,
  _now: number
): MemoryFlowDecision {
  const pressure = computeMemoryPressure(node);

  // Low pressure: hold everything
  if (pressure < 50) {
    return { action: 'hold' };
  }

  // Medium pressure: hold if high-value
  if (pressure < 80) {
    if (item.informationDensity >= 0.7) {
      return { action: 'hold' };
    }
  }

  // High pressure: must decide
  if (pressure >= 80) {
    // Low-value items: vent
    if (item.informationDensity < 0.5) {
      const heatReleased = Math.ceil(item.heatDebt);
      return { action: 'vent', heatReleased };
    }

    // High-value items: try routing
    const lowPressureNeighbor = findLowPressureNeighbor(node);
    if (lowPressureNeighbor && pressureGradient(node, lowPressureNeighbor) > 0) {
      return {
        action: 'route',
        toNode: lowPressureNeighbor as LiquidMemoryNode<unknown>,
      };
    }

    // No lower-pressure neighbor: vent as fallback
    const heatReleased = Math.ceil(item.heatDebt);
    return { action: 'vent', heatReleased };
  }

  // Fallback: hold
  return { action: 'hold' };
}

// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
// Venting (Landauer Erasure)
// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

/**
 * Vent an item from a node (Landauer erasure).
 *
 * Removes the item and returns the heat released (accumulated heatDebt).
 * Increases node's ventRate and decreases held memory.
 */
export function ventItem<T>(
  node: LiquidMemoryNode<T>,
  key: string
): number {
  const item = node.items.get(key);
  if (!item) return 0;

  const heatReleased = item.heatDebt;

  // Update budget
  node.budget.held = Math.max(0, node.budget.held - 1);
  node.budget.ventRate += 1;

  // Remove item
  node.items.delete(key);

  return heatReleased;
}

/**
 * Vent the lowest-value items until pressure drops below threshold.
 *
 * Returns total heat released.
 */
export function ventUntilPressure<T>(
  node: LiquidMemoryNode<T>,
  targetPressure: number
): number {
  let totalHeatReleased = 0;

  while (computeMemoryPressure(node) > targetPressure && node.items.size > 0) {
    // Find lowest-density item
    let worstKey: string | null = null;
    let worstDensity = 1.0;

    for (const [key, item] of node.items.entries()) {
      if (item.informationDensity < worstDensity) {
        worstKey = key;
        worstDensity = item.informationDensity;
      }
    }

    if (!worstKey) break;

    const heat = ventItem(node, worstKey);
    totalHeatReleased += heat;
  }

  return totalHeatReleased;
}

// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
// Routing (Mesh Flow)
// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

/**
 * Route an item from one node to another (down pressure gradient).
 *
 * Precondition: toNode should have lower pressure than fromNode.
 * Updates budgets and moves item via callback (caller handles persistence).
 */
export function routeItem<T>(
  fromNode: LiquidMemoryNode<T>,
  key: string,
  toNode: LiquidMemoryNode<T>
): boolean {
  const item = fromNode.items.get(key);
  if (!item) return false;

  // Update source budget
  fromNode.budget.held = Math.max(0, fromNode.budget.held - 1);
  fromNode.budget.ventRate += 1; // Routing counts as "controlled vent"

  // Update destination budget
  toNode.budget.held += 1;
  toNode.budget.ventRate = Math.max(0, toNode.budget.ventRate - 1); // Reduce vent as we fill

  // Move item
  fromNode.items.delete(key);
  toNode.items.set(key, item);

  return true;
}

// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
// Consolidation (Sleep-Like Recovery)
// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

/**
 * Consolidation: compress low-fitness items to reduce overall cost.
 *
 * Like sleepConsolidation in MemoryAsRetrocausalLoan.lean:
 * Reduce informationDensity of low-value items, releasing heat without erasure.
 * Useful during idle periods to strengthen the remaining memory.
 */
export function consolidateNode<T>(
  node: LiquidMemoryNode<T>,
  fitnessScore: (item: LiquidItem<T>) => number
): number {
  let totalHeatReleased = 0;

  for (const item of node.items.values()) {
    const fitness = fitnessScore(item);

    // Low-fitness items: reduce density and heat debt
    if (fitness < 0.3) {
      const density_reduction = Math.max(0, item.informationDensity - 0.2);
      const heat_released = Math.ceil(item.heatDebt * density_reduction);

      item.informationDensity = Math.max(0, item.informationDensity - 0.2);
      item.heatDebt = Math.max(0, item.heatDebt - heat_released);

      totalHeatReleased += heat_released;
    }
  }

  return totalHeatReleased;
}

// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
// Node Creation and Management
// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

/**
 * Create a new liquid memory node with initial budget.
 */
export function createLiquidMemoryNode<T>(
  id: string,
  capacity: number,
  initialVentRate: number = 0,
  neighbors: () => LiquidMemoryNode<T>[] = () => []
): LiquidMemoryNode<T> {
  return {
    id,
    budget: {
      capacity,
      held: memHeld(capacity, initialVentRate),
      ventRate: initialVentRate,
    },
    items: new Map(),
    neighbors: neighbors as any,
  };
}

/**
 * Check if a node can hold a new item (space available).
 */
export function canHoldItem<T>(node: LiquidMemoryNode<T>): boolean {
  return node.items.size < node.budget.held;
}

/**
 * Attempt to add an item to a node.
 * Returns true if added, false if node is full and venting/routing didn't help.
 */
export function addItemWithFlowDecision<T>(
  node: LiquidMemoryNode<T>,
  item: LiquidItem<T>,
  now: number
): boolean {
  // If space available, add directly
  if (canHoldItem(node)) {
    node.items.set(item.key, item);
    return true;
  }

  // Node is full; try venting low-value items
  const pressure = computeMemoryPressure(node);
  if (pressure > 80) {
    ventUntilPressure(node, 70);
  }

  // Try again
  if (canHoldItem(node)) {
    node.items.set(item.key, item);
    return true;
  }

  // If still full, try routing to neighbors
  for (const neighbor of node.neighbors()) {
    if (neighbor.items.size < neighbor.budget.held) {
      node.items.set(item.key, item);
      routeItem(node, item.key, neighbor);
      return true;
    }
  }

  // Overflow: venting new item is a last resort (indicates budget exceeded)
  return false;
}

/**
 * System-wide consolidation sweep across all nodes.
 *
 * Called periodically (e.g., during idle time) to strengthen memory.
 */
export function consolidateAllNodes<T>(
  nodes: LiquidMemoryNode<T>[],
  fitnessScore: (item: LiquidItem<T>) => number
): number {
  let totalHeat = 0;
  for (const node of nodes) {
    totalHeat += consolidateNode(node, fitnessScore);
  }
  return totalHeat;
}
