// @a0n/gnosis/src/liveness.ts
/**
 * Liveness — Helix visualization + ejection plans from checkpoint homology
 *
 * Combines organism topology (per-project health from real code topology)
 * with the checkpoint homology algebra to produce:
 *   1. Helix nodes — visualization data with real sigma (code stress colors)
 *   2. Ejection plans — what to do with each unhealthy project
 *   3. God Formula repair ordering — which functions to fix first
 *
 * The helix visualization maps sigma to color temperature:
 *   - sigma ~0: cool (healthy, learning)
 *   - sigma ~0.5: warm (stressed, inflamed)
 *   - sigma ~1: hot (cancerous, necrotic)
 *
 * @module @a0n/gnosis/liveness
 */

import type {
  OrganismNode,
  OrganismTopology,
} from './organism-topology';
import type {
  ProjectCheckpointProfile,
  RepairCandidate,
} from './polyglot-checkpoint-bridge';
import type {
  CheckpointedSystem,
  HealthStatus,
  RepairPath,
} from './checkpoint-substrates';
import {
  forecastRepair,
  forecastNeglect,
  diagnose,
  godFormulaWeight,
} from './checkpoint-substrates';

// ═══════════════════════════════════════════════════════════════════════
// Helix Visualization
// ═══════════════════════════════════════════════════════════════════════

/**
 * A node in the liveness helix visualization.
 */
export interface HelixNode {
  /** Project name. */
  name: string;
  /** Relative path from workspace root. */
  path: string;
  /** Real sigma from checkpoint homology (0 = healthy, 1 = dead). */
  sigma: number;
  /** Health classification. */
  health: HealthStatus;
  /** Repair path recommendation. */
  repairPath: RepairPath;
  /** Color temperature for visualization (0 = cool blue, 1 = hot red). */
  colorTemperature: number;
  /** Helix angle (radians) — position on the spiral. */
  angle: number;
  /** Helix radius — proportional to function count. */
  radius: number;
  /** Helix z-position — ordered by sigma. */
  z: number;
  /** Whether the project is alive (has consumers / dependents). */
  isAlive: boolean;
  /** Active channel count (functions with beta1 > 0). */
  activeChannels: number;
  /** Destroyed channel count (functions with beta1 = 0). */
  destroyedChannels: number;
}

/**
 * Full helix visualization data.
 */
export interface HelixData {
  /** Helix nodes sorted by sigma (healthiest at top). */
  nodes: HelixNode[];
  /** Workspace-level stats. */
  stats: {
    totalProjects: number;
    aliveProjects: number;
    deadProjects: number;
    avgSigma: number;
    maxSigma: number;
  };
}

/**
 * Compute helix visualization data from organism topology.
 *
 * Sigma comes directly from the CheckpointedSystem's sicknessFraction —
 * computed from real beta1 and deficit, not estimates.
 *
 * @param topology - Organism topology with per-project profiles
 * @param aliveProjects - Set of project names known to be alive (have consumers)
 */
export function computeHelixData(
  topology: OrganismTopology,
  aliveProjects?: Set<string>,
): HelixData {
  const alive = aliveProjects ?? new Set(topology.nodes.map((n) => n.name));

  // Sort by sigma ascending (healthiest first = top of helix)
  const sorted = [...topology.nodes].sort((a, b) => a.sigma - b.sigma);

  const nodes: HelixNode[] = sorted.map((node, i) => {
    const isAlive = alive.has(node.name);

    // Base sigma from real code topology
    let sigma = node.sigma;
    // Boost sigma for dead/orphaned projects (additive — same topology, worse prognosis)
    if (!isAlive) sigma = Math.min(1.0, sigma + 0.2);

    // Color temperature: maps sigma to [0, 1] visual range
    // Healthy (sigma ~0) → cool blue (0)
    // Stressed (sigma ~0.3) → warm yellow (0.5)
    // Cancerous (sigma ~0.8) → hot red (1.0)
    const colorTemperature = Math.min(1.0, sigma * 1.25);

    // Helix geometry: golden angle for even distribution
    const goldenAngle = Math.PI * (3 - Math.sqrt(5));
    const angle = i * goldenAngle;
    const maxFuncs = Math.max(
      1,
      ...topology.nodes.map((n) => n.profile?.scan.functionCount ?? 1),
    );
    const funcCount = node.profile?.scan.functionCount ?? 1;
    const radius = 0.3 + 0.7 * (funcCount / maxFuncs);
    const z = sorted.length > 1 ? i / (sorted.length - 1) : 0.5;

    return {
      name: node.name,
      path: node.relativePath,
      sigma,
      health: node.health,
      repairPath: node.profile?.diagnosis.repairPath ?? 'monitoring',
      colorTemperature,
      angle,
      radius,
      z,
      isAlive,
      activeChannels: node.profile?.activeChannels ?? 0,
      destroyedChannels: node.profile?.destroyedChannels ?? 0,
    };
  });

  const sigmas = nodes.map((n) => n.sigma);
  const totalSigma = sigmas.reduce((s, v) => s + v, 0);

  return {
    nodes,
    stats: {
      totalProjects: nodes.length,
      aliveProjects: nodes.filter((n) => n.isAlive).length,
      deadProjects: nodes.filter((n) => !n.isAlive).length,
      avgSigma: nodes.length > 0 ? totalSigma / nodes.length : 0,
      maxSigma: sigmas.length > 0 ? Math.max(...sigmas) : 0,
    },
  };
}

// ═══════════════════════════════════════════════════════════════════════
// Ejection Plans
// ═══════════════════════════════════════════════════════════════════════

/** Strategy for an ejection plan. */
export type EjectionStrategy =
  | 'monitor'           // healthy — just watch
  | 'adopt-project'     // stressed/inflamed — restore channels
  | 'exorcise-project'  // cancerous — remove entirely
  | 'necropsy';         // necrotic — document and archive

/**
 * An ejection plan for an unhealthy project.
 */
export interface EjectionPlan {
  /** Target project name. */
  target: string;
  /** Target project path. */
  targetPath: string;
  /** Ejection strategy based on diagnosis. */
  strategy: EjectionStrategy;
  /** Health status that triggered the plan. */
  health: HealthStatus;
  /** Sigma at diagnosis time. */
  sigma: number;
  /** Repair path from checkpoint algebra. */
  repairPath: RepairPath;
  /** Human-readable rationale. */
  rationale: string;
  /**
   * Top functions to repair first (God Formula ordering).
   * Only present for adopt-project strategy.
   */
  repairOrdering: RepairCandidate[];
  /**
   * Forecast: system state if top candidate is repaired.
   */
  forecastIfRepaired: CheckpointedSystem | null;
  /**
   * Forecast: system state if neglected (lose one more channel).
   */
  forecastIfNeglected: CheckpointedSystem | null;
}

/**
 * Generate ejection plans for all projects in an organism topology.
 *
 * Every project gets a plan — healthy ones get 'monitor', unhealthy
 * ones get actionable strategies with God Formula repair ordering.
 */
export function generateEjectionPlans(
  topology: OrganismTopology,
  aliveProjects?: Set<string>,
): EjectionPlan[] {
  const alive = aliveProjects ?? new Set(topology.nodes.map((n) => n.name));

  return topology.nodes.map((node) => {
    const isAlive = alive.has(node.name);
    const profile = node.profile;

    const strategy = classifyStrategy(node.health, isAlive);
    const rationale = buildRationale(node, isAlive, strategy);

    // God Formula repair ordering for adopt-project strategy
    const repairOrdering: RepairCandidate[] =
      strategy === 'adopt-project' && profile
        ? profile.repairOrdering.slice(0, 5)
        : [];

    return {
      target: node.name,
      targetPath: node.relativePath,
      strategy,
      health: node.health,
      sigma: node.sigma,
      repairPath: profile?.diagnosis.repairPath ?? 'monitoring',
      rationale,
      repairOrdering,
      forecastIfRepaired: profile?.forecastIfRepaired ?? null,
      forecastIfNeglected: profile?.forecastIfNeglected ?? null,
    };
  });
}

function classifyStrategy(
  health: HealthStatus,
  isAlive: boolean,
): EjectionStrategy {
  if (health === 'necrotic') return 'necropsy';
  if (health === 'cancerous') return isAlive ? 'adopt-project' : 'exorcise-project';
  if (health === 'inflamed' || health === 'stressed') return 'adopt-project';
  return 'monitor';
}

function buildRationale(
  node: OrganismNode,
  isAlive: boolean,
  strategy: EjectionStrategy,
): string {
  const profile = node.profile;
  const sigma = node.sigma.toFixed(3);

  switch (strategy) {
    case 'monitor':
      return `${node.name} is healthy (sigma=${sigma}). Continue monitoring.`;

    case 'adopt-project': {
      const destroyed = profile?.destroyedChannels ?? 0;
      const total = (profile?.activeChannels ?? 0) + destroyed;
      const top = profile?.repairOrdering[0];
      const hint = top
        ? ` Start with ${top.functionName} (godWeight=${top.godWeight}).`
        : '';
      return (
        `${node.name} has ${destroyed}/${total} deaf functions (sigma=${sigma}). ` +
        `Adopt and restore beta1 via integration.${hint}`
      );
    }

    case 'exorcise-project':
      return (
        `${node.name} is cancerous (sigma=${sigma}) and has no alive consumers. ` +
        `Exorcise: remove from workspace or archive.`
      );

    case 'necropsy':
      return (
        `${node.name} is necrotic (sigma=${sigma}, all channels destroyed). ` +
        `Document the failure mode and archive.`
      );
  }
}

// ═══════════════════════════════════════════════════════════════════════
// Reporting
// ═══════════════════════════════════════════════════════════════════════

/**
 * Format ejection plans as a text report.
 */
export function formatEjectionReport(plans: EjectionPlan[]): string {
  const lines: string[] = [];

  lines.push('Ejection Plans');
  lines.push('='.repeat(60));

  const byStrategy = {
    'exorcise-project': plans.filter((p) => p.strategy === 'exorcise-project'),
    necropsy: plans.filter((p) => p.strategy === 'necropsy'),
    'adopt-project': plans.filter((p) => p.strategy === 'adopt-project'),
    monitor: plans.filter((p) => p.strategy === 'monitor'),
  };

  for (const [strategy, group] of Object.entries(byStrategy)) {
    if (group.length === 0) continue;

    lines.push('');
    lines.push(`--- ${strategy.toUpperCase()} (${group.length}) ---`);

    for (const plan of group) {
      lines.push(`  ${plan.target} [sigma=${plan.sigma.toFixed(3)}]`);
      lines.push(`    ${plan.rationale}`);

      if (plan.repairOrdering.length > 0) {
        lines.push('    Fix first:');
        for (const candidate of plan.repairOrdering) {
          lines.push(
            `      - ${candidate.functionName} in ${candidate.filePath} ` +
              `(godWeight=${candidate.godWeight}, beta1=${candidate.beta1})`,
          );
        }
      }

      if (plan.forecastIfRepaired) {
        const after = plan.forecastIfRepaired;
        lines.push(
          `    If repaired: sigma -> ${after.sicknessFraction.toFixed(3)} ` +
            `(beta1=${after.beta1}, deficit=${after.deficit})`,
        );
      }
      if (plan.forecastIfNeglected) {
        const after = plan.forecastIfNeglected;
        lines.push(
          `    If neglected: sigma -> ${after.sicknessFraction.toFixed(3)} ` +
            `(beta1=${after.beta1}, deficit=${after.deficit})`,
        );
      }
    }
  }

  return lines.join('\n');
}

/**
 * Format helix data as a text report.
 */
export function formatHelixReport(helix: HelixData): string {
  const lines: string[] = [];

  lines.push('Liveness Helix');
  lines.push('='.repeat(60));
  lines.push(
    `Projects: ${helix.stats.totalProjects} | ` +
      `Alive: ${helix.stats.aliveProjects} | ` +
      `Dead: ${helix.stats.deadProjects} | ` +
      `Avg sigma: ${helix.stats.avgSigma.toFixed(3)} | ` +
      `Max sigma: ${helix.stats.maxSigma.toFixed(3)}`,
  );
  lines.push('');

  for (const node of helix.nodes) {
    const temp = node.colorTemperature;
    const bar = '#'.repeat(Math.round(temp * 20)).padEnd(20, '.');
    lines.push(
      `  ${node.name.padEnd(35)} sigma=${node.sigma.toFixed(3)} ` +
        `[${bar}] ${node.health} (${node.repairPath})`,
    );
  }

  return lines.join('\n');
}
