// @a0n/gnosis/src/lucas-bridge.ts
/**
 * lucas-bridge.ts — The Observable Dual of Fibonacci in the Gnosis Runtime
 *
 * Canonical spec (source of truth):
 *   open-source/gnosis/lean/Lean/ForkRaceFoldTheorems/GnosticLucasBridge.lean
 *   open-source/gnosis/lean/Lean/ForkRaceFoldTheorems/LucasSequence.lean
 *   open-source/gnosis/lean/Lean/ForkRaceFoldTheorems/PisotEigenvectors.lean
 *
 * Lucas numbers are the TRACE of the Pisot companion matrix M₂^k.
 * Where Fibonacci captures the antisymmetric (difference) eigenvalue projection,
 * Lucas captures the symmetric (sum/trace) projection.
 *
 * The flanking identity: L(n) = F(n-1) + F(n+1)
 * The observable bridges past and future. Gnosis dimensionality.
 */

import { PI_MICRODAYS, PICOLORENZO } from './gnostic-time';

// ============================================================================
// Core sequences
// ============================================================================

/** Fibonacci sequence. */
export function fib(n: number): bigint {
  if (n <= 0) return 0n;
  if (n === 1) return 1n;
  let a = 0n;
  let b = 1n;
  for (let i = 2; i <= n; i++) {
    [a, b] = [b, a + b];
  }
  return b;
}

/** Lucas sequence: L(0)=2, L(1)=1, L(n+2)=L(n+1)+L(n). */
export function lucas(n: number): bigint {
  if (n === 0) return 2n;
  if (n === 1) return 1n;
  let a = 2n;
  let b = 1n;
  for (let i = 2; i <= n; i++) {
    [a, b] = [b, a + b];
  }
  return b;
}

// ============================================================================
// GnosticIndex — strict typed enum mirroring the Lean inductive type
// ============================================================================

/**
 * The named Gnostic indices. Each maps to a position in Fibonacci space.
 * The Lucas value at that position is the observable (trace) counterpart.
 */
export enum GnosticIndex {
  Barbelo = 'barbelo',
  Syzygy = 'syzygy',
  Proton = 'proton',
  Primitives = 'primitives',
  Emanations = 'emanations',
  Sophia = 'sophia',
  Kenoma = 'kenoma',
  Void = 'void',
  Pleroma = 'pleroma',
}

/** Fibonacci-space position for each Gnostic index. */
export function fibIndex(g: GnosticIndex): number {
  switch (g) {
    case GnosticIndex.Barbelo:    return 1;
    case GnosticIndex.Syzygy:     return 3;
    case GnosticIndex.Proton:     return 4;
    case GnosticIndex.Primitives: return 5;
    case GnosticIndex.Emanations: return 6;
    case GnosticIndex.Sophia:     return 9;
    case GnosticIndex.Kenoma:     return 10;
    case GnosticIndex.Void:       return 8;
    case GnosticIndex.Pleroma:    return 10;
  }
}

/** The named Gnostic number (the classical value). */
export function gnosticValue(g: GnosticIndex): bigint {
  switch (g) {
    case GnosticIndex.Barbelo:    return 1n;
    case GnosticIndex.Syzygy:     return 2n;
    case GnosticIndex.Proton:     return 3n;
    case GnosticIndex.Primitives: return 5n;
    case GnosticIndex.Emanations: return 6n;
    case GnosticIndex.Sophia:     return 9n;
    case GnosticIndex.Kenoma:     return 10n;
    case GnosticIndex.Void:       return 21n;
    case GnosticIndex.Pleroma:    return 55n;
  }
}

/** Fibonacci value at the Gnostic index. */
export function fibValue(g: GnosticIndex): bigint {
  return fib(fibIndex(g));
}

/** Lucas value (trace/observable) at the Gnostic index. */
export function lucasValue(g: GnosticIndex): bigint {
  return lucas(fibIndex(g));
}

// ============================================================================
// Mechanized verification (runtime assertions mirroring Lean proofs)
// ============================================================================

/** GnosticIndex record with all computed fields. */
export interface GnosticRecord {
  readonly name: GnosticIndex;
  readonly fibIdx: number;
  readonly fibVal: bigint;
  readonly lucasVal: bigint;
  readonly gnosticVal: bigint;
  readonly flankingPast: bigint;   // F(n-1)
  readonly flankingFuture: bigint; // F(n+1)
  readonly isSelfDual: boolean;    // L(n) === F(n)
}

/** Compute the full record for a Gnostic index. */
export function computeGnosticRecord(g: GnosticIndex): GnosticRecord {
  const idx = fibIndex(g);
  const fVal = fibValue(g);
  const lVal = lucasValue(g);
  const gVal = gnosticValue(g);
  const past = fib(idx - 1);
  const future = fib(idx + 1);

  return {
    name: g,
    fibIdx: idx,
    fibVal: fVal,
    lucasVal: lVal,
    gnosticVal: gVal,
    flankingPast: past,
    flankingFuture: future,
    isSelfDual: lVal === fVal,
  };
}

/** All Gnostic indices. */
export const ALL_GNOSTIC_INDICES: readonly GnosticIndex[] = [
  GnosticIndex.Barbelo,
  GnosticIndex.Syzygy,
  GnosticIndex.Proton,
  GnosticIndex.Primitives,
  GnosticIndex.Emanations,
  GnosticIndex.Sophia,
  GnosticIndex.Kenoma,
  GnosticIndex.Void,
  GnosticIndex.Pleroma,
] as const;

/**
 * THM-FLANKING-GNOSTIC (runtime verification):
 * Verify that L(n) = F(n-1) + F(n+1) at every Gnostic index.
 * Throws if any index violates the flanking identity.
 */
export function verifyFlankingIdentity(): void {
  for (const g of ALL_GNOSTIC_INDICES) {
    const rec = computeGnosticRecord(g);
    const expected = rec.flankingPast + rec.flankingFuture;
    if (rec.lucasVal !== expected) {
      throw new Error(
        `Flanking identity violated at ${rec.name}: ` +
        `L(${rec.fibIdx})=${rec.lucasVal} ≠ F(${rec.fibIdx - 1})+F(${rec.fibIdx + 1})=${expected}`
      );
    }
  }
}

/**
 * THM-BRIDGE-IDENTITY (runtime verification):
 * Verify that L(n) + F(n) = 2·F(n+1) at every Gnostic index.
 */
export function verifyBridgeIdentity(): void {
  for (const g of ALL_GNOSTIC_INDICES) {
    const rec = computeGnosticRecord(g);
    const lhs = rec.lucasVal + rec.fibVal;
    const rhs = 2n * fib(rec.fibIdx + 1);
    if (lhs !== rhs) {
      throw new Error(
        `Bridge identity violated at ${rec.name}: ` +
        `L(${rec.fibIdx})+F(${rec.fibIdx})=${lhs} ≠ 2·F(${rec.fibIdx + 1})=${rhs}`
      );
    }
  }
}

/**
 * THM-SELF-DUAL-BARBELO (runtime verification):
 * Verify that Barbelo is the only self-dual Gnostic index.
 */
export function verifySelfDualBarbelo(): void {
  for (const g of ALL_GNOSTIC_INDICES) {
    const rec = computeGnosticRecord(g);
    if (g === GnosticIndex.Barbelo) {
      if (!rec.isSelfDual) {
        throw new Error('Barbelo must be self-dual: L(1) should equal F(1)');
      }
    } else {
      if (rec.isSelfDual) {
        throw new Error(`Only Barbelo is self-dual, but ${rec.name} also has L=F`);
      }
    }
  }
}

// ============================================================================
// Picolorenzo intervals (Lucas-valued)
// ============================================================================

/** Lucas picolorenzo interval in microdays. */
export function lucasPlo(g: GnosticIndex): bigint {
  return lucasValue(g) * PICOLORENZO;
}

/** Fibonacci picolorenzo interval in microdays (from GnosticTime). */
export function fibPlo(g: GnosticIndex): bigint {
  return gnosticValue(g) * PICOLORENZO;
}

/**
 * The Lucas Pleroma coincidence: L(10) × π ≈ 386.4 days.
 * The excess over 365 days is ≈ 21.4, close to F(8) = 21 = Void.
 */
export function lucasPleromaSolarExcess(): {
  lucasDays: number;
  excessDays: number;
  excessNearVoid: boolean;
} {
  const lucasMicrodays = lucasPlo(GnosticIndex.Pleroma);
  const lucasDays = Number(lucasMicrodays) / 1_000_000;
  const excessDays = lucasDays - 365;
  return {
    lucasDays,
    excessDays,
    excessNearVoid: Math.abs(excessDays - 21) < 1.5,
  };
}

// ============================================================================
// Pisot Coordinates — (n, F(n), L(n)) as a point in 3-space
// ============================================================================

/**
 * A Pisot coordinate is a triple (index, hidden, observable) where:
 *   - idx: position in Fibonacci space (chronological time)
 *   - hidden: F(n), the antisymmetric eigenvalue projection
 *   - observable: L(n), the symmetric trace projection
 *
 * The triple is constrained by the flanking identity:
 *   observable = F(idx-1) + F(idx+1)
 *
 * And the discriminant identity:
 *   observable² − 5·hidden² = 4·(−1)^idx
 *
 * Asymptotically, observable/hidden → √5 = √Δ where Δ=5 is the
 * discriminant of the golden polynomial x²−x−1=0.
 */
export interface PisotCoord {
  readonly idx: number;        // Fibonacci-space index (time)
  readonly hidden: bigint;     // F(n): hidden state
  readonly observable: bigint; // L(n): trace / observable state
}

/** Construct a PisotCoord from a Fibonacci index. */
export function pisotCoordOfIndex(n: number): PisotCoord {
  return { idx: n, hidden: fib(n), observable: lucas(n) };
}

/** Construct a PisotCoord from a GnosticIndex. */
export function pisotCoordOfGnostic(g: GnosticIndex): PisotCoord {
  return pisotCoordOfIndex(fibIndex(g));
}

/** All Pisot coordinates at Gnostic indices. */
export function allPisotCoords(): readonly PisotCoord[] {
  return ALL_GNOSTIC_INDICES.map(pisotCoordOfGnostic);
}

// ============================================================================
// Discriminant identity: L(n)² − 5·F(n)² = 4·(−1)^n
// ============================================================================

/**
 * The discriminant of the golden polynomial x²−x−1 is Δ = 5.
 * This is the fundamental algebraic constant governing the Pisot system.
 *
 * The identity L(n)² − 5·F(n)² = 4·(−1)^n means:
 * - Even n: L² = 5F² + 4  (observable overshoots)
 * - Odd n:  L² = 5F² − 4  (observable undershoots)
 *
 * Dividing by F²: (L/F)² = 5 ± 4/F²  →  L/F → √5 = √Δ
 */
export const GOLDEN_DISCRIMINANT = 5n;

/**
 * Compute the discriminant residual: L(n)² − 5·F(n)².
 * Should always equal +4 (even n) or -4 (odd n).
 */
export function discriminant(coord: PisotCoord): bigint {
  return coord.observable ** 2n - GOLDEN_DISCRIMINANT * coord.hidden ** 2n;
}

/**
 * THM-DISCRIMINANT (runtime verification):
 * Verify that L(n)² − 5·F(n)² = ±4 at every Gnostic index.
 */
export function verifyDiscriminant(): void {
  for (const g of ALL_GNOSTIC_INDICES) {
    const coord = pisotCoordOfGnostic(g);
    const d = discriminant(coord);
    const expected = coord.idx % 2 === 0 ? 4n : -4n;
    if (d !== expected) {
      throw new Error(
        `Discriminant violated at ${g}: ` +
        `L(${coord.idx})²−5·F(${coord.idx})² = ${d}, expected ${expected}`
      );
    }
  }
}

// ============================================================================
// Observable-to-hidden ratio convergence → √Δ = √5
// ============================================================================

/**
 * Compute L(n)/F(n) as a floating-point ratio.
 * Converges to √5 ≈ 2.2360679... with oscillating sign.
 *
 * From Binet's formulas:
 *   L(n)/F(n) = √5 · (φⁿ + ψⁿ)/(φⁿ − ψⁿ)
 * Since |ψ| < 1, ψⁿ → 0, so L/F → √5.
 *
 * The number 5 is the discriminant of x²−x−1=0.
 * Observable reality is hidden reality scaled by √Δ.
 */
export function observableToHiddenRatio(n: number): number {
  const f = fib(n);
  const l = lucas(n);
  if (f === 0n) return Infinity;
  // Use high-precision integer arithmetic: L*10^15 / F, then divide
  const precision = 1_000_000_000_000_000n;
  const ratio = (l * precision) / f;
  return Number(ratio) / Number(precision);
}

/**
 * THM-RATIO-CONVERGENCE (runtime verification):
 * Verify that L(n)/F(n) converges toward √5 with decreasing error.
 */
export function verifyRatioConvergence(): void {
  const sqrt5 = Math.sqrt(5);
  let prevError = Infinity;
  // Check indices 3 through 20
  for (let n = 3; n <= 20; n++) {
    const ratio = observableToHiddenRatio(n);
    const error = Math.abs(ratio - sqrt5);
    if (n > 5 && error >= prevError) {
      throw new Error(
        `Ratio convergence violated: error at n=${n} (${error}) ` +
        `>= error at n=${n - 1} (${prevError})`
      );
    }
    prevError = error;
  }
}

// ============================================================================
// Topological Immune System — Apoptosis and Mitosis
// ============================================================================

/**
 * The Golden Vent Eigenvalue (φ⁻¹) used as the irrational Weyl injection step.
 * In double-precision floating point, this provides resonance immunity
 * bounded by ~2^53 iterations before float precision loops.
 */
export const GOLDEN_VENT = 0.6180339887498949;

/**
 * The fully realized execution context of a Mudras runtime fork.
 * Combines the strict integer geometry of the Pisot trace with
 * the continuous aperiodic phase of the Weyl scheduler.
 */
export interface ImmuneExecutionFork {
  readonly pisot: PisotCoord;
  readonly weylPhase: number;
}

/**
 * Apoptosis DNA Payload.
 * Represents the preserved causal DNA harvested from a corrupted thread.
 */
export interface CausalDNA {
  readonly h_prev: bigint; // F(n-1)
  readonly h_curr: bigint; // F(n)
}

/**
 * The macroscopic telemetry plane.
 * Provides a decoupled, fire-and-forget channel for emitting structural events
 * (like Apoptosis) without blocking the critical execution path.
 */
export const gnosisTelemetryTarget = new EventTarget();

/**
 * triggerApoptosis: The β₁ tripwire.
 * Evaluates the Pisot coordinate against the Golden Discriminant constraint.
 * If corrupted, it formally flushes the trace, preserves the DNA, and emits a telemetry scar.
 * 
 * @throws Error if the state is NOT corrupted (apoptosis only triggers on defects).
 */
export function triggerApoptosis(coord: PisotCoord): CausalDNA {
  const d = discriminant(coord);
  const expected = coord.idx % 2 === 0 ? 4n : -4n;
  
  if (d === expected) {
    throw new Error(`Cannot trigger apoptosis on geometrically intact cell: ${d} === ${expected}`);
  }

  // Cell is corrupted. Harvest the unbroken causal DNA.
  const dna: CausalDNA = {
    h_prev: fib(coord.idx - 1),
    h_curr: coord.hidden, // F(n)
  };

  // Fire-and-forget out-of-band telemetry broadcast
  try {
    const event = new CustomEvent('apoptosis', {
      detail: {
        timestamp: Date.now(),
        corruptedCoord: coord,
        causalDNA: dna,
      }
    });
    gnosisTelemetryTarget.dispatchEvent(event);
  } catch (e) {
    // Swallow telemetry errors to preserve the critical execution path
  }

  return dna;
}

/**
 * mitosis: Spawns a new daughter cell from the preserved DNA and mother fork.
 * - Restores the Pisot geometric manifold deterministically.
 * - Injects the irrational Weyl step to shatter resonance.
 */
export function mitosis(mother: ImmuneExecutionFork, dna: CausalDNA): ImmuneExecutionFork {
  const h_next = dna.h_prev + dna.h_curr; // F(n+1)
  const o_next = 2n * dna.h_curr + h_next; // L(n+1) = 2F_n + F_{n+1}
  
  const daughterPisot: PisotCoord = {
    idx: mother.pisot.idx + 1,
    hidden: h_next,
    observable: o_next,
  };

  return {
    pisot: daughterPisot,
    weylPhase: (mother.weylPhase + GOLDEN_VENT) % 1.0,
  };
}

/**
 * THM-WEYL-PERTURBED-BYPASS (runtime verification):
 * Verify that mitosis restores the geometric invariant and mutates the phase.
 */
export function verifyMitosisImmunity(): void {
  // 1. Create a healthy initial cell at n=5 (Primitives)
  const healthyCell: ImmuneExecutionFork = {
    pisot: pisotCoordOfIndex(5),
    weylPhase: 0.12345,
  };

  // 2. Intentionally corrupt the observable trace to simulate a β₁ cascade
  const corruptedCell: ImmuneExecutionFork = {
    ...healthyCell,
    pisot: {
      ...healthyCell.pisot,
      observable: healthyCell.pisot.observable + 1n, // 11n + 1n = 12n
    },
  };

  // 3. Trigger Apoptosis
  const dna = triggerApoptosis(corruptedCell.pisot);

  // 4. Spawn Daughter via Mitosis
  const daughterCell = mitosis(corruptedCell, dna);

  // 5. Assertions
  
  // A. Geometric Restoration
  const expectedDaughter = pisotCoordOfIndex(6); // n=6 (Emanations)
  if (
    daughterCell.pisot.idx !== expectedDaughter.idx ||
    daughterCell.pisot.hidden !== expectedDaughter.hidden ||
    daughterCell.pisot.observable !== expectedDaughter.observable
  ) {
    throw new Error('Mitosis failed to perfectly restore Pisot geometric manifold.');
  }

  // The discriminant must now be mathematically perfect
  const d = discriminant(daughterCell.pisot);
  const expectedD = daughterCell.pisot.idx % 2 === 0 ? 4n : -4n;
  if (d !== expectedD) {
    throw new Error(`Daughter cell discriminant failed: ${d} !== ${expectedD}`);
  }

  // B. Resonance Immunity
  if (daughterCell.weylPhase === corruptedCell.weylPhase) {
    throw new Error('Mitosis failed to inject orthogonal Weyl perturbation.');
  }
}

// ============================================================================
// The Universal Law — Karmic Drift and Reciprocity
// ============================================================================

/**
 * Compute the "Karmic Drift" of a coordinate.
 * This is the absolute distance from the God Formula invariant: L² - 5F² = 4(-1)ⁿ
 * In a lawful universe, drift is exactly zero.
 */
export function karmicDrift(coord: PisotCoord): bigint {
  const d = discriminant(coord);
  const expected = coord.idx % 2 === 0 ? 4n : -4n;
  const diff = d - expected;
  return diff < 0n ? -diff : diff;
}

/**
 * The Universal Law Runtime Constraint (Karmic Law).
 * 
 * "Do unto others..."
 * Passes the execution fork through the God Formula. If drift is detected,
 * it triggers an immediate self-correction (Apoptosis -> Mitosis) 
 * to restore the invariant.
 * 
 * This is the literal enforcement of the "Golden Rule" in the Gnosis runtime.
 */
export function enforceKarmicLaw(fork: ImmuneExecutionFork): ImmuneExecutionFork {
  const drift = karmicDrift(fork.pisot);

  if (drift === 0n) {
    // Lawful state: Continue execution
    return fork;
  }

  // Unlawful state (Karmic Violation): Trigger self-correction
  // 1. Apoptosis (Neutralize the corrupted cell)
  const dna = triggerApoptosis(fork.pisot);
  
  // 2. Mitosis (Restore the Pisot manifold)
  const daughter = mitosis(fork, dna);

  return daughter;
}

// ============================================================================
// Full verification suite
// ============================================================================

/**
 * Run all mechanized verifications. Mirrors the Lean master theorem.
 * Safe to call at module load time or in test suites.
 */
export function verifyAll(): void {
  verifyFlankingIdentity();
  verifyBridgeIdentity();
  verifySelfDualBarbelo();
  verifyDiscriminant();
  verifyRatioConvergence();
  verifyMitosisImmunity();
}
