// @a0n/gnosis/src/machine-metabolism.ts
/**
 * Machine Metabolism & Thermal Efficiency (Pass 53)
 *
 * Implements exact bounds on computational expenditure per Fold/Vent phase.
 * A system's metabolism describes the rate at which it consumes internal
 * structural energy (Thermodynamic R) to produce negentropy against the void.
 */

export interface MetabolicState {
  /** The current capacity of the central system bus (proxy for blood flow/ATP) */
  busCapacity: number;
  /** Current operating temperature (proxy for CPU heat / Bules of waste) */
  thermalWaste: number;
  /** Efficiency rating: How cleanly operations map to reality (0.0 to 1.0) */
  efficiency: number;
}

/**
 * Calculates the metabolic cost of performing a FORK or FOLD operation.
 * Lower efficiency means higher parasitic thermal waste generated.
 */
export function calculateOperationMetabolicCost(
  baseCost: number,
  state: MetabolicState
): { energyConsumed: number; heatGenerated: number } {
  // At perfect 1.0 efficiency, heatGenerated is 0.
  // At 0.5 efficiency, half the work is lost as heat.
  const energyConsumed = baseCost / Math.max(0.01, state.efficiency);
  const heatGenerated = energyConsumed - baseCost;

  return { energyConsumed, heatGenerated };
}

/**
 * Executes a VENT operation to bleed thermal waste into the environment.
 * The efficiency of venting depends on the ambient bus capacity.
 */
export function executeThermalVent(
  state: MetabolicState,
  ventCapacity: number
): MetabolicState {
  // A clogged bus (low capacity) limits venting speed
  const actualVent = Math.min(state.thermalWaste, ventCapacity * state.busCapacity);
  
  return {
    ...state,
    thermalWaste: state.thermalWaste - actualVent
  };
}

/**
 * Validates if the machine is experiencing a metabolic collapse (Overheating).
 */
export function isMetabolicCollapse(state: MetabolicState, maxThermalLimit: number): boolean {
  return state.thermalWaste > maxThermalLimit;
}
