// @a0n/gnosis/src/math-sandbox.ts
/**
 * **Not verified by Lean.** Anchors and optional deterministic numeric toys for demos.
 *
 * For formal content see `lean/Lean/MathSandbox.lean` and `docs/MATH_SANDBOX.md`.
 * `ZeckendorfFST` is Init-only (see `docs/ZECKENDORF_FST_PEEL.md`). `PisotVickreyTable` dropped
 * `Mathlib.Tactic.Ring` (see `docs/PISOT_VICKREY_PEEL.md`); `GreekLogicCanon` monolith remains separate.
 *
 * @module @a0n/gnosis/math-sandbox
 */

/** Stable Lean library paths for tooling, logs, and cross-repo trace anchors. */
export const GNOSIS_MATH_SANDBOX_LEAN_MODULES = [
  'MathSandbox',
  'ForkRaceFoldTheorems.GnosisMath.Basic',
  'ForkRaceFoldTheorems.GnosisMath.Fibonacci',
  'ForkRaceFoldTheorems.GnosisMath.ListNat',
  'ForkRaceFoldTheorems.GnosisMathPrelude',
  'ForkRaceFoldTheorems.GreekLogicCanon.DiscreteBoundary',
  'ForkRaceFoldTheorems.KraftInequality',
  'ForkRaceFoldTheorems.GeneralizedParity',
  'ForkRaceFoldTheorems.BaseTenConsequence',
  'ForkRaceFoldTheorems.LeibnizCalculemus',
  'ForkRaceFoldTheorems.Hyperoperations',
  'MathSandboxExtras',
] as const;

export type GnosisMathSandboxLeanModule =
  (typeof GNOSIS_MATH_SANDBOX_LEAN_MODULES)[number];

/**
 * Modules documented for refactor before full Mathlib-reduction integration.
 * See `docs/MATH_SANDBOX.md` — especially GreekLogicCanon fork plan.
 */
export const GNOSIS_MATH_SANDBOX_INTEGRATION_CANDIDATES = [
  'ForkRaceFoldTheorems.GreekLogicCanon',
  'ForkRaceFoldTheorems.PisotVickreyTable',
] as const;

export type GnosisMathSandboxIntegrationCandidate =
  (typeof GNOSIS_MATH_SANDBOX_INTEGRATION_CANDIDATES)[number];

/** Lean theorem family name for cross-linking numeric demos (not a proof obligation in TS). */
export const KRAFT_ADDRESS_SPACE_LEAN_MODULE = 'ForkRaceFoldTheorems.KraftInequality';

/** Explicit disclaimer string for manifests and UIs. */
export const MATH_SANDBOX_SCRIPT_DISCLAIMER =
  'This TypeScript module is not proof-checked against the Lean math sandbox; use `pnpm run validate:lean-minimal` (or `lake build MathSandbox` in heritage `lean/`) for the Init-first graph.';

/**
 * Discrete Kraft sum for alphabet size `D` and codeword lengths (toy check, not verified vs Lean).
 * Returns Σᵢ D^(-lᵢ) as a floating-point value.
 */
export function kraftSumFloat(D: number, lengths: readonly number[]): number {
  if (D <= 1 || !Number.isFinite(D)) {
    throw new RangeError('kraftSumFloat: D must be a finite number > 1');
  }
  let s = 0;
  for (const len of lengths) {
    if (len < 1 || !Number.isFinite(len)) {
      throw new RangeError('kraftSumFloat: each length must be a finite number ≥ 1');
    }
    s += D ** -len;
  }
  return s;
}

/**
 * Zeckendorf-style Fibonacci weights for demos: `fibZ(0)…fibZ(count−1)` matching Lean
 * `GnosisMath.fibZ` / `ZeckendorfFST.F` (`1, 1, 2, 3, 5, …`). Not proof-checked in TS.
 *
 * @param count how many weights to return (≥ 0); use `0` for `[]`
 */
export function fibZWeights(count: number): readonly number[] {
  if (!Number.isInteger(count) || count < 0) {
    throw new RangeError('fibZWeights: count must be a non-negative integer');
  }
  if (count === 0) {
    return [];
  }
  if (count === 1) {
    return [1];
  }
  const out: number[] = [1, 1];
  for (let i = 2; i < count; i++) {
    out.push(out[i - 1]! + out[i - 2]!);
  }
  return out;
}
