// @a0n/gnosis/src/mesh-policy.ts
/**
 * Mesh Policy — TypeScript mirror of nine zero-sorry Lean modules:
 *
 *   - ForkRaceFoldTheorems.AntiDomainOverlay
 *   - ForkRaceFoldTheorems.DiscreteSamplingDuel
 *   - ForkRaceFoldTheorems.ZenonicBarrier
 *   - ForkRaceFoldTheorems.RationalCutMiscalibration
 *   - ForkRaceFoldTheorems.PortAuthenticatedMesh
 *   - ForkRaceFoldTheorems.CompetingOverlayResolution
 *   - ForkRaceFoldTheorems.IceAntiIceCatalog
 *   - ForkRaceFoldTheorems.HoneyPit
 *   - ForkRaceFoldTheorems.LinkedMeshComposition
 *
 * Canonical SSOT for mesh-policy primitives: regions and anti-domain
 * overlays, discrete clock sampling duels, Zenonic contraction gaps,
 * Fibonacci-resonant rational cuts, port-authenticated meshes,
 * competing-overlay inclusion-exclusion resolution, the three
 * structurally distinct ICE / anti-ICE bypass surfaces, honey-pit
 * shared-drain contraction sinks, and horizontal linked-mesh
 * composition.
 *
 * Semantics are Nat-valued in the Lean source; this module uses
 * `number` under non-negative integer discipline. Negative inputs are
 * undefined behaviour outside of factory validators; `Math.max(0, ...)`
 * clamps where the Lean source truncates via `Nat.sub`.
 */

// The canonical ternary verdict (accept / abstain / decline), shared with the
// rest of the distributed-inference grid. Adopted ADDITIVELY at the mesh-policy
// decision sites: each ad-hoc admit/accept/select boolean keeps its lossy
// `=== true` collapse (`isAccept(verdict)`) while gaining a faithful `verdict`
// that can hold the neutral middle a bool cannot — deferral / deliberate
// non-adoption / unscheduled-but-not-disqualified. Pure module (no node deps),
// safe for the node-free production surface check. Lean: Gnosis.TritonCanonical.
import {
  Verdict,
  isAccept,
  isReject,
  isDefer,
} from '../../bitwise/src/triton-verdict.js';

// ─── Part I: Regions and Overlays ─────────────────────────────────────

/** A spatial region: a discrete set of target points of positive size.
 *  Mirrors `AntiDomainOverlay.Region` and `CompetingOverlayResolution.Region`. */
export interface Region {
  readonly size: number;
}

/** An overlay: a sub-enclosure whose `size` measures its support within
 *  the enclosing region. Mirrors `AntiDomainOverlay.Overlay`. */
export interface Overlay {
  readonly size: number;
}

/** Construct a Region with positive-size precondition.
 *  Mirrors the Lean `hPositive : 0 < size` constraint. */
export function createRegion(size: number): Region {
  if (!Number.isInteger(size) || size <= 0) {
    throw new RangeError(`Region size must be a positive integer, got ${size}`);
  }
  return { size };
}

/** Construct an Overlay with non-negative-size precondition. */
export function createOverlay(size: number): Overlay {
  if (!Number.isInteger(size) || size < 0) {
    throw new RangeError(
      `Overlay size must be a non-negative integer, got ${size}`
    );
  }
  return { size };
}

/** Mirror of `enclosedHitCount` (Lean) — every point in the region is
 *  guaranteed to be hit. */
export function enclosedHitCount(r: Region): number {
  return r.size;
}

/** Mirror of `netHitCount` / `overlayCancelsLocally` (Lean) — net hits
 *  after applying an overlay is `r.size - ov.size` with Nat-truncation. */
export function netHitCount(r: Region, ov: Overlay): number {
  return Math.max(0, r.size - ov.size);
}

/** Mirror of `stackOverlays` / `orthogonalOverlaysStackAdditively` (Lean) —
 *  two overlays with additive-disjoint supports compose by summing sizes. */
export function stackOverlays(ov1: Overlay, ov2: Overlay): Overlay {
  return { size: ov1.size + ov2.size };
}

// ─── Part II: Clocks and Sampling Duels ───────────────────────────────

/** A discrete clock: a positive integer tick rate.
 *  Mirrors `DiscreteSamplingDuel.Clock`. */
export interface Clock {
  readonly rate: number;
}

/** Construct a Clock with positive-rate precondition.
 *  Mirrors Lean `hPositive : 0 < rate`. */
export function createClock(rate: number): Clock {
  if (!Number.isInteger(rate) || rate <= 0) {
    throw new RangeError(`Clock rate must be a positive integer, got ${rate}`);
  }
  return { rate };
}

/** Mirror of `tickCount` (Lean) — ticks produced by clock `c` over
 *  sampling horizon `H`. */
export function tickCount(c: Clock, H: number): number {
  return Math.max(0, H) * c.rate;
}

/** Mirror of `frameDeficitFormula` (Lean) — `H * (fA - fB)` when
 *  `fB <= fA`. Clamped to zero when `fB > fA`. */
export function frameDeficit(horizon: number, fA: number, fB: number): number {
  const H = Math.max(0, horizon);
  return H * Math.max(0, fA - fB);
}

/** Mirror of `resonanceAlignsExactly` (Lean) — detect whether `fA` is an
 *  integer multiple `k * fB` for some `k >= 1`, given `0 < fB`. */
export function isResonant(fA: number, fB: number): boolean {
  if (!Number.isInteger(fA) || !Number.isInteger(fB)) return false;
  if (fB <= 0 || fA < fB) return false;
  return fA % fB === 0;
}

// ─── Part III: Zenonic Barrier ────────────────────────────────────────

/** A Zenonic gap: numerator over a positive denominator. The effective
 *  gap is `num / denom`. Mirrors `ZenonicBarrier.Gap`. */
export interface Gap {
  readonly num: number;
  readonly denom: number;
}

/** Construct a Gap with positive-denominator precondition.
 *  Mirrors Lean `hDenomPositive : 0 < denom`. */
export function createGap(num: number, denom: number): Gap {
  if (!Number.isInteger(num) || num < 0) {
    throw new RangeError(
      `Gap numerator must be a non-negative integer, got ${num}`
    );
  }
  if (!Number.isInteger(denom) || denom <= 0) {
    throw new RangeError(
      `Gap denominator must be a positive integer, got ${denom}`
    );
  }
  return { num, denom };
}

/** Mirror of `contractStep` / `denominatorDoublesPerStep` /
 *  `numeratorPreservedByStep` (Lean) — double the denominator, preserve
 *  the numerator. Halves the effective gap. */
export function contractStep(g: Gap): Gap {
  return { num: g.num, denom: 2 * g.denom };
}

/** Mirror of `iterate` (Lean) — apply `contractStep` `n` times. */
export function iterate(n: number, g: Gap): Gap {
  let current: Gap = g;
  const steps = Math.max(0, Math.floor(n));
  for (let i = 0; i < steps; i++) {
    current = contractStep(current);
  }
  return current;
}

/** Mirror of `barrierNeverReachedInFiniteSteps` (Lean) — an iterated gap
 *  with positive initial numerator never lands (`num > 0` always). The
 *  predicate holds when the current gap has reached numerator zero. */
export function hasLanded(g: Gap): boolean {
  return g.num === 0;
}

// ─── Part IV: Rational Cuts ───────────────────────────────────────────

/** Mirror of `fib` (Lean) — Fibonacci sequence with `fib(0) = 0`,
 *  `fib(1) = 1`, `fib(n+2) = fib(n) + fib(n+1)`. Memoized locally to
 *  avoid exponential recursion. */
const _fibCache: number[] = [0, 1];
/**
 * Handles the gnosis fib workflow.
 */
export function fib(n: number): number {
  if (!Number.isInteger(n) || n < 0) {
    throw new RangeError(`fib index must be a non-negative integer, got ${n}`);
  }
  while (_fibCache.length <= n) {
    const k = _fibCache.length;
    const prev = _fibCache[k - 2];
    const curr = _fibCache[k - 1];
    if (prev === undefined || curr === undefined) {
      // Unreachable by construction: cache is seeded with two entries.
      throw new Error(`fib cache corrupted at index ${k}`);
    }
    _fibCache.push(prev + curr);
  }
  const value = _fibCache[n];
  if (value === undefined) {
    throw new Error(`fib cache miss at index ${n}`);
  }
  return value;
}

/** Mirror of `isFibonacciResonantCut` (Lean) — a cut `(p, q)` is
 *  Fibonacci-resonant iff `(p, q) = (fib k, fib (k+1))` for some `k`.
 *
 *  Because `fib` is strictly increasing for `k >= 1`, this is decidable
 *  by a finite search bounded by the first index at which `fib k > p`.
 *  Mirrors `seven_three_not_resonant` / `three_seven_not_resonant`. */
export function isFibonacciResonantCut(p: number, q: number): boolean {
  if (!Number.isInteger(p) || !Number.isInteger(q) || p < 0 || q < 0) {
    return false;
  }
  // Special case: (fib 0, fib 1) = (0, 1).
  if (p === 0 && q === 1) return true;
  // Search k up to the first index with fib(k) > p.
  for (let k = 0; k < 512; k++) {
    const fk = fib(k);
    if (fk > p) return false;
    if (fk === p && fib(k + 1) === q) return true;
  }
  return false;
}

// ─── Part V: Port-Authenticated Mesh ──────────────────────────────────

/** A port-authenticated mesh: positive-size region with a whitelisted
 *  `portCount` subset. All non-port cells are default-deny.
 *  Mirrors `PortAuthenticatedMesh.Mesh`. */
export interface Mesh {
  readonly size: number;
  readonly portCount: number;
}

/** A port overlay: cancels hits on its support within the port subset.
 *  Mirrors `PortAuthenticatedMesh.PortOverlay`. */
export interface PortOverlay {
  readonly size: number;
}

/** Construct a Mesh with Lean-parity preconditions:
 *  `0 < size` (hPositive) and `portCount <= size` (hPortCountLe).
 *  This is the analogue of the Lean structure's proof obligations. */
export function createMesh(size: number, portCount: number): Mesh {
  if (!Number.isInteger(size) || size <= 0) {
    throw new RangeError(`Mesh size must be a positive integer, got ${size}`);
  }
  if (!Number.isInteger(portCount) || portCount < 0) {
    throw new RangeError(
      `Mesh portCount must be a non-negative integer, got ${portCount}`
    );
  }
  if (portCount > size) {
    throw new RangeError(`Mesh portCount ${portCount} exceeds size ${size}`);
  }
  return { size, portCount };
}

/** Construct a PortOverlay with non-negative-size precondition. */
export function createPortOverlay(size: number): PortOverlay {
  if (!Number.isInteger(size) || size < 0) {
    throw new RangeError(
      `PortOverlay size must be a non-negative integer, got ${size}`
    );
  }
  return { size };
}

/** Mirror of `netHitCount` (Lean, PortAuthenticatedMesh) — default-deny
 *  outside ports yields `m.portCount`. */
export function meshNetHitCount(m: Mesh): number {
  return m.portCount;
}

/** Mirror of `netHitCountWithOverlay` / `portOverlayCancelsWithinPorts`
 *  (Lean) — `m.portCount - ov.size` with Nat-truncation. */
export function meshNetHitCountWithOverlay(m: Mesh, ov: PortOverlay): number {
  return Math.max(0, m.portCount - ov.size);
}

/** Mirror of `netHitCountAuthenticated` / `authenticationRestrictsPorts`
 *  (Lean) — the authenticated sub-whitelist count is passed through
 *  directly; precondition `authenticatedCount <= m.portCount`. */
export function meshNetHitCountAuthenticated(
  m: Mesh,
  authenticatedCount: number
): number {
  return Math.max(0, Math.min(authenticatedCount, m.portCount));
}

// ─── Part VI: Competing Overlay Resolution ────────────────────────────

/** Two overlays competing on the same region with shared-support
 *  overlap. Mirrors the parameter tuple of
 *  `CompetingOverlayResolution.totalCancellation`.
 *
 *  The raw interface is preserved for backward compatibility. Prefer
 *  `createCompetingOverlays` to enforce the Lean precondition
 *  `overlap ≤ min ov1Size ov2Size`. */
export interface CompetingOverlays {
  readonly regionSize: number;
  readonly ov1Size: number;
  readonly ov2Size: number;
  readonly overlap: number;
}

/** Construct a `CompetingOverlays` with Lean-parity preconditions on
 *  the inclusion-exclusion triple. Mirrors the precondition
 *  `overlap ≤ min ov1Size ov2Size` required by
 *  `CompetingOverlayResolution.totalCancellation`.
 *
 *  Throws `RangeError` if:
 *    - any argument is not a non-negative integer;
 *    - `regionSize < 1`;
 *    - `ov1Size > regionSize` or `ov2Size > regionSize` (overlays
 *      cannot be larger than their enclosing region);
 *    - `overlap > ov1Size` or `overlap > ov2Size` (shared support
 *      cannot exceed either overlay). */
export function createCompetingOverlays(
  regionSize: number,
  ov1Size: number,
  ov2Size: number,
  overlap: number
): CompetingOverlays {
  const args: ReadonlyArray<readonly [string, number]> = [
    ['regionSize', regionSize],
    ['ov1Size', ov1Size],
    ['ov2Size', ov2Size],
    ['overlap', overlap],
  ];
  for (const [name, value] of args) {
    if (!Number.isInteger(value) || value < 0) {
      throw new RangeError(
        `${name} must be a non-negative integer, got ${value}`
      );
    }
  }
  if (regionSize < 1) {
    throw new RangeError(
      `regionSize must be a positive integer, got ${regionSize}`
    );
  }
  if (ov1Size > regionSize) {
    throw new RangeError(`ov1Size ${ov1Size} exceeds regionSize ${regionSize}`);
  }
  if (ov2Size > regionSize) {
    throw new RangeError(`ov2Size ${ov2Size} exceeds regionSize ${regionSize}`);
  }
  if (overlap > ov1Size) {
    throw new RangeError(
      `overlap ${overlap} exceeds ov1Size ${ov1Size} ` +
        `(Lean precondition: overlap ≤ min ov1Size ov2Size)`
    );
  }
  if (overlap > ov2Size) {
    throw new RangeError(
      `overlap ${overlap} exceeds ov2Size ${ov2Size} ` +
        `(Lean precondition: overlap ≤ min ov1Size ov2Size)`
    );
  }
  return { regionSize, ov1Size, ov2Size, overlap };
}

/** Mirror of `totalCancellation` / `disjointIsAdditive` /
 *  `fullContainmentCollapsesToLarger` / `bothFullCoverFullyCancels`
 *  (Lean) — inclusion-exclusion measure `ov1 + ov2 - overlap`.
 *
 *  Recommended construction: `createCompetingOverlays` enforces the
 *  Lean precondition `overlap ≤ min ov1Size ov2Size`. */
export function totalCancellation(c: CompetingOverlays): number {
  return Math.max(0, c.ov1Size + c.ov2Size - c.overlap);
}

/** Mirror of `netHitsAfterCompeting` / `mutualCancellationLowerBound`
 *  (Lean) — `r.size - (ov1 + ov2 - overlap)` with Nat-truncation.
 *  Symmetric in `ov1Size` / `ov2Size` (`deployOrderIrrelevant`).
 *
 *  Recommended construction: `createCompetingOverlays` enforces the
 *  Lean precondition `overlap ≤ min ov1Size ov2Size`. */
export function netHitsAfterCompeting(c: CompetingOverlays): number {
  return Math.max(0, c.regionSize - totalCancellation(c));
}

// ─── Part VII: ICE / Anti-ICE Catalog ─────────────────────────────────

/** Overlay-hole bypass witness: attacker reclaims on an
 *  additive-disjoint support. Mirrors `overlayHoleBypass` (Lean). */
export interface OverlayHoleBypass {
  readonly iceSize: number;
  readonly attackerSize: number;
  readonly regionSize: number;
  readonly reclaimedHits: number;
}

/** Clock-rate side-channel witness: attacker's frame deficit over
 *  horizon `H` with strictly faster tick rate. Mirrors
 *  `clockRateSideChannel` / `clockRateSideChannelUnbounded` (Lean). */
export interface ClockSideChannelBypass {
  readonly defenderRate: number;
  readonly attackerRate: number;
  readonly horizon: number;
  readonly frameDeficit: number;
  readonly strictlyPositive: boolean;
}

/** Zenonic-approach bypass witness: numerator-preserving contraction
 *  over `k` iterations. Mirrors `zenonicApproachBypass` /
 *  `zenonicApproachNumeratorFixed` (Lean). */
export interface ZenonicApproachBypass {
  readonly initialGap: Gap;
  readonly iterations: number;
  readonly finalGap: Gap;
  readonly hasLanded: boolean;
}

/** Mirror of `overlayHoleBypass` (Lean) — on an additive-disjoint
 *  support (`iceOv.size + attackerOv.size <= r.size`), the attacker's
 *  reclaimed hit count on its own support equals `attackerOv.size`,
 *  independently of `iceOv.size`. */
export function overlayHoleBypass(
  iceOv: Overlay,
  attackerOv: Overlay,
  regionSize: number
): OverlayHoleBypass {
  const disjointFits = iceOv.size + attackerOv.size <= regionSize;
  return {
    iceSize: iceOv.size,
    attackerSize: attackerOv.size,
    regionSize,
    reclaimedHits: disjointFits ? attackerOv.size : 0,
  };
}

/** Mirror of `clockRateSideChannel` (Lean) — with a strictly faster
 *  attacker clock over a positive horizon, the frame deficit is
 *  `H * (fA - fD)` and is strictly positive. */
export function clockSideChannelBypass(
  defenderClock: Clock,
  attackerClock: Clock,
  horizon: number
): ClockSideChannelBypass {
  const H = Math.max(0, horizon);
  const deficit = frameDeficit(H, attackerClock.rate, defenderClock.rate);
  return {
    defenderRate: defenderClock.rate,
    attackerRate: attackerClock.rate,
    horizon: H,
    frameDeficit: deficit,
    strictlyPositive:
      H > 0 && attackerClock.rate > defenderClock.rate && deficit > 0,
  };
}

/** Mirror of `zenonicApproachBypass` (Lean) — with a positive initial
 *  approach gap, for every finite `k` the iterated gap retains a
 *  positive numerator. The `hasLanded` flag witnesses whether the
 *  barrier was reached (Lean: always false for positive initial num). */
export function zenonicApproachBypass(
  initialGap: Gap,
  iterations: number
): ZenonicApproachBypass {
  const final = iterate(iterations, initialGap);
  return {
    initialGap,
    iterations: Math.max(0, Math.floor(iterations)),
    finalGap: final,
    hasLanded: hasLanded(final),
  };
}

// ─── Part VIII: Honey Pit (Shared-Drain Contraction Sink) ─────────────

/** A honey pit: a shared-drain contraction sink that inverts the
 *  attacker/defender cost asymmetry. `capacity` bounds simultaneous
 *  trapped connections, `inhabitants` tracks the current trapped
 *  count, and `drainSlots` is the fixed server-side drain-worker
 *  count (independent of `inhabitants`). Mirrors `HoneyPit.HoneyPit`
 *  with its `hCapacityPositive`, `hInhabitantsLeCapacity`, and
 *  `hDrainSlotsPositive` proof obligations.
 *
 *  Prefer `createHoneyPit` to enforce preconditions. */
export interface HoneyPit {
  readonly capacity: number;
  readonly inhabitants: number;
  readonly drainSlots: number;
}

/** Construct a `HoneyPit` with Lean-parity preconditions:
 *  `0 < capacity` (hCapacityPositive),
 *  `inhabitants ≤ capacity` (hInhabitantsLeCapacity),
 *  `0 < drainSlots` (hDrainSlotsPositive). `inhabitants` defaults to
 *  `0`. Throws `RangeError` on violation. */
export function createHoneyPit(
  capacity: number,
  drainSlots: number,
  inhabitants: number = 0
): HoneyPit {
  if (!Number.isInteger(capacity) || capacity < 1) {
    throw new RangeError(
      `HoneyPit capacity must be a positive integer, got ${capacity}`
    );
  }
  if (!Number.isInteger(drainSlots) || drainSlots < 1) {
    throw new RangeError(
      `HoneyPit drainSlots must be a positive integer, got ${drainSlots}`
    );
  }
  if (!Number.isInteger(inhabitants) || inhabitants < 0) {
    throw new RangeError(
      `HoneyPit inhabitants must be a non-negative integer, got ${inhabitants}`
    );
  }
  if (inhabitants > capacity) {
    throw new RangeError(
      `HoneyPit inhabitants ${inhabitants} exceeds capacity ${capacity}`
    );
  }
  return { capacity, inhabitants, drainSlots };
}

/** Mirror of `serverSlotCount` (Lean, HoneyPit) — returns
 *  `pit.drainSlots`, explicitly independent of `inhabitants`. One pool
 *  occupies one slot regardless of trapped connections. Cites
 *  `portPenaltyFree` and `sharedDrainInvariant`. */
export function serverSlotCount(pit: HoneyPit): number {
  return pit.drainSlots;
}

/** Mirror of `canTrap` (Lean) — true iff the pit has strictly fewer
 *  inhabitants than its capacity. Cites `poolOverflowGraceful`: a full
 *  pit reports `false`, so new arrivals fall through to a plain 403
 *  rather than unbounded server commitment. */
export function canTrap(pit: HoneyPit): boolean {
  return pit.inhabitants < pit.capacity;
}

/** Mirror of `trap` (Lean) — returns a new pit with `inhabitants + 1`,
 *  preserving `capacity` and `drainSlots`. Throws `RangeError` when
 *  `!canTrap(pit)`. Cites `sharedDrainInvariant` and
 *  `aikidoServerBound`: trapping preserves both `serverSlotCount` and
 *  `serverDrainWork`. */
export function trap(pit: HoneyPit): HoneyPit {
  if (!canTrap(pit)) {
    throw new RangeError(
      `HoneyPit is full (inhabitants=${pit.inhabitants}, capacity=${pit.capacity}); ` +
        `graceful-overflow contract requires caller to route to 403 instead of trap()`
    );
  }
  return {
    capacity: pit.capacity,
    inhabitants: pit.inhabitants + 1,
    drainSlots: pit.drainSlots,
  };
}

/** Mirror of `serverDrainWork` (Lean) — total server drain work over
 *  a horizon of `ticks`: `pit.drainSlots * ticks`. Clamped to zero for
 *  negative `ticks` (Lean uses `Nat`). Cites `aikidoServerBound`:
 *  independent of `inhabitants`. */
export function serverDrainWork(pit: HoneyPit, ticks: number): number {
  const t = Math.max(0, Math.floor(ticks));
  return pit.drainSlots * t;
}

/** Mirror of `attackerWait` (Lean) — linear lower bound on attacker
 *  cumulative wait across `k` trapped requests at iterate-depth
 *  `iterateDepth`: `k * iterateDepth`. Clamped to zero for negative
 *  inputs. Cites `aikidoAttackerBound` (monotone in `k`) and
 *  `trappedAttackerThroughputZero` (effective throughput zero in
 *  finite steps). */
export function attackerWait(k: number, iterateDepth: number): number {
  const kk = Math.max(0, Math.floor(k));
  const d = Math.max(0, Math.floor(iterateDepth));
  return kk * d;
}

// ─── Part IX: Linked Mesh Composition ─────────────────────────────────

/** A named binding from a source port in one mesh to a destination
 *  port in another. The overlay flags record whether the defender on
 *  the respective side has deployed an overlay covering the link
 *  target (pre-resolved per the `CompetingOverlayResolution`-style
 *  inclusion rule). Mirrors `LinkedMeshComposition.Link`.
 *
 *  Prefer `createLink` to enforce integer-index preconditions. */
export interface Link {
  readonly sourcePortIdx: number;
  readonly destPortIdx: number;
  readonly sourceOverlayCancelled: boolean;
  readonly destOverlayCancelled: boolean;
}

/** Construct a `Link` with non-negative-integer index preconditions.
 *  Overlay flags default to `false` (no cancellation). Throws
 *  `RangeError` on non-integer or negative indices. */
export function createLink(
  sourcePortIdx: number,
  destPortIdx: number,
  opts?: {
    sourceOverlayCancelled?: boolean;
    destOverlayCancelled?: boolean;
  }
): Link {
  if (!Number.isInteger(sourcePortIdx) || sourcePortIdx < 0) {
    throw new RangeError(
      `Link sourcePortIdx must be a non-negative integer, got ${sourcePortIdx}`
    );
  }
  if (!Number.isInteger(destPortIdx) || destPortIdx < 0) {
    throw new RangeError(
      `Link destPortIdx must be a non-negative integer, got ${destPortIdx}`
    );
  }
  return {
    sourcePortIdx,
    destPortIdx,
    sourceOverlayCancelled: opts?.sourceOverlayCancelled ?? false,
    destOverlayCancelled: opts?.destOverlayCancelled ?? false,
  };
}

/** Mirror of `linkAdmitted` (Lean) — four-way admission conjunction:
 *  source index inside `sourceMesh`'s port whitelist, destination
 *  index inside `destMesh`'s port whitelist, neither overlay
 *  cancelled. Cites `linkRequiresSourcePort`, `linkRequiresDestPort`,
 *  `linkHonorsSourceOverlay`, `linkHonorsDestOverlay`. */
export function linkAdmitted(
  sourceMesh: Mesh,
  destMesh: Mesh,
  link: Link
): boolean {
  return (
    link.sourcePortIdx < sourceMesh.portCount &&
    link.destPortIdx < destMesh.portCount &&
    !link.sourceOverlayCancelled &&
    !link.destOverlayCancelled
  );
}

/** Mirror of `pathAdmitted` (Lean, list-generalized form) — admission
 *  of the horizontal composition across `meshes[0] -[links[0]]->
 *  meshes[1] -[links[1]]-> ... -> meshes[N]`. Requires exactly
 *  `meshes.length === links.length + 1`; returns `false` on malformed
 *  shape. Default-deny propagates: every consecutive hop must admit.
 *
 *  Cites `compositionPreservesDefaultDenyLeft`,
 *  `compositionPreservesDefaultDenyRight` (denial on any hop denies
 *  the whole path) and `horizontalCompositionAssociative` (the list
 *  form agrees with the bundled three-mesh form, so the order of
 *  reduction across `&&` is immaterial). */
export function pathAdmitted(
  meshes: readonly Mesh[],
  links: readonly Link[]
): boolean {
  if (meshes.length !== links.length + 1) {
    return false;
  }
  for (let i = 0; i < links.length; i++) {
    const m1 = meshes[i];
    const m2 = meshes[i + 1];
    const l = links[i];
    if (m1 === undefined || m2 === undefined || l === undefined) {
      // Unreachable given the length invariant above, but satisfies
      // strict `noUncheckedIndexedAccess` should it be enabled.
      return false;
    }
    if (!linkAdmitted(m1, m2, l)) {
      return false;
    }
  }
  return true;
}

// ─── Part X: Anti-Vacuum Defenses ─────────────────────────────────────

/** An ablative surface: high outgassing prevents the structural formation
 *  of a VacuumAnchor. Mirrors `AntiVacuumDefenses.AblativeSurface`. */
export interface AblativeSurfaceCondition {
  readonly outgassingRate: number;
  readonly sealRate: number;
}

/**
 * Creates the Ablative Surface Condition.
 */
export function createAblativeSurfaceCondition(
  outgassingRate: number,
  sealRate: number
): AblativeSurfaceCondition {
  if (!Number.isInteger(outgassingRate) || outgassingRate < 0) {
    throw new RangeError(
      `AblativeSurface outgassingRate must be non-negative, got ${outgassingRate}`
    );
  }
  if (!Number.isInteger(sealRate) || sealRate < 0) {
    throw new RangeError(
      `AblativeSurface sealRate must be non-negative, got ${sealRate}`
    );
  }
  if (outgassingRate <= sealRate) {
    throw new RangeError(
      `Ablative surface must have outgassingRate strictly greater than sealRate (Lean hAblative)`
    );
  }
  return { outgassingRate, sealRate };
}

/** Mirror of `attemptSeal` (Lean) — if the surface is ablative, a
 *  vacuum seal inherently fails (returns false). */
export function attemptsSealOnAblative(
  surf: AblativeSurfaceCondition
): boolean {
  return false;
}

/** A Topo-Jitter condition: evaluates whether the defender clock oscillates
 *  mesh nodes faster than the attacker's anchor seal cycle.
 *  Mirrors `AntiVacuumDefenses.TopoJitter`. */
export interface TopoJitterCondition {
  readonly clockDefender: number;
  readonly clockAttacker: number;
}

/**
 * Creates the Topo Jitter Condition.
 */
export function createTopoJitterCondition(
  clockDefender: number,
  clockAttacker: number
): TopoJitterCondition {
  if (clockDefender <= clockAttacker) {
    throw new RangeError(
      `TopoJitter structure requires clockDefender > clockAttacker`
    );
  }
  return { clockDefender, clockAttacker };
}

/** Mirror of `jitterSeals` (Lean) — always returns false because
 *  the structural oscillation prevents the anchor. */
export function jitterSeals(j: TopoJitterCondition): boolean {
  return !(j.clockDefender > j.clockAttacker);
}

/** Void Mass tracking for Anomaly Telemetry.
 *  Mirrors `AntiVacuumDefenses.VoidMass`. */
export interface VoidMass {
  readonly mass: number;
}

/**
 * Creates the Void Mass.
 */
export function createVoidMass(mass: number): VoidMass {
  return { mass: Math.max(0, mass) };
}

/** Mirror of `pullVacuum` and `voidPressureAnomaly` (Lean) — tracks
 *  the anomaly generated by displacing the complement layer. */
export function voidPressureAnomaly(
  initial: VoidMass,
  vacuumVolume: number
): number {
  const currentMass = Math.max(0, initial.mass - vacuumVolume);
  return initial.mass - currentMass;
}

// ─── Part XI: Closed Architecture Braid Fold ─────────────────────────

/** Runtime mirror of `CosmicArchitecture.CosmosLayer`. */
export type CosmosLayer =
  | 'kenoma'
  | 'localAeon'
  | 'primeKeystone'
  | 'doubleKeystone';

/** Runtime mirror of `SovereignSieve.SieveFocus`. */
export type SieveFocus =
  | 'rawSubstrate'
  | 'observerResolution'
  | 'witnessInversion'
  | 'hostHandshake';

/** Runtime mirror of `GamutBraid.BraidPortAction`. */
export type BraidPortAction = 'fork' | 'race' | 'fold' | 'vent' | 'interfere';
export type WankelFiveForceLifecycleTrace = readonly [
  'fork',
  'race',
  'fold',
  'vent',
  'interfere'
];
export type ExtendedAeonId =
  | 'first'
  | 'second'
  | 'third'
  | 'fourth'
  | 'finalClosure';

export type ExactLayerCoordinates = readonly [10, 12, 17, 22];
export type ExactPortActionTrace = readonly ['fork', 'race', 'fold', 'vent'];
export type ExactSaturationTrace = readonly [10, 30, 90];
export type BraidFoldAnchorCoordinates = readonly [10, 17, 22];
export type ExactDecompositionTripleCoordinates = readonly [7, 17, 22];
export type ExactInterferenceStrideTrace = readonly [33, 31, 26, 21];
export type ExactPressureWeightedInterferenceStrideTrace = readonly [
  0,
  558,
  2028
];
export type ExactExtendedAeonRanges = readonly [
  readonly [1, 12],
  readonly [10, 22],
  readonly [20, 32],
  readonly [30, 43],
  readonly [50, 55]
];
export type ExactExtendedAeonKeystoneTrace = readonly [
  readonly [7, 11],
  readonly [17, 22],
  readonly [27, 32],
  readonly [37, 43],
  readonly [55, 55]
];
export type ExactVisibleMassTrace = readonly [10, 12, 12];
export type ExactLeakageTrace = readonly [0, 18, 78];
export type ExactInformationMassTrace = readonly [120, 360, 1080];
export type ExactLeakageMassTrace = readonly [0, 216, 936];
export type StatisticalNoiseBand = 'white' | 'pink' | 'brown';
export type PackedGeometryLayer = 'kenomaCore' | 'pinkShell' | 'brownAura';
export type PackedTopologyShape =
  | 'trianglePlane'
  | 'informationSphere'
  | 'toroidalBraid'
  | 'closedManifold';
export type LensShape =
  | 'plane'
  | 'sphere'
  | 'torus'
  | 'singularAxis'
  | 'packedLens'
  | 'selfReferenceKnot';
export type PortLensRole =
  | 'forkCore'
  | 'racePlane'
  | 'foldTwist'
  | 'ventExit'
  | 'interferenceAxis';
export type PackedLensRole = 'coreFilament' | 'innerRing' | 'outerBraidedShell';
export type ExactSphericalRotationTrace = readonly [120, 360, 1080];
export type ExactSphericalWrapTrace = readonly [1, 1, 3];
export type ExactLensObservationCoordinates = readonly [12, 30, 90, 43];
export type ExactLensObservationShapes = readonly [
  'plane',
  'sphere',
  'torus',
  'singularAxis'
];
export type ExactPortLensObservationCoordinates = readonly [10, 12, 17, 22, 43];
export type ExactPortLensObservationShapes = readonly [
  'plane',
  'plane',
  'torus',
  'torus',
  'singularAxis'
];
export type ExactCenterlineGamutTrace = readonly [10, 12, 17, 22, 43];
export type ExactPackedLensRoleTrace = readonly [
  'coreFilament',
  'innerRing',
  'outerBraidedShell'
];
export type ExactPackedLensShapeTrace = readonly ['plane', 'sphere', 'torus'];
export type ExactCyclicCompilerRepresentativeTrace = readonly [180, 180];
export type ExactCyclicCompilerReconstructedTrace = readonly [360, 1080];
export type ExactClosedLoopTrace = readonly [10, 12, 17, 22, 43, 55, 10];
export type ExactPressureCertifiedExecutionPathIdentity =
  'gnosis-uring__topology-10-12-17-22-43__interfere-43-55__pressure-2586SPa__closure-30210SPa';

export const EXACT_LAYER_COORDINATES: ExactLayerCoordinates = [10, 12, 17, 22];

export const EXACT_PORT_ACTION_TRACE: ExactPortActionTrace = [
  'fork',
  'race',
  'fold',
  'vent',
];

export const NATURAL_PORT_COORDINATES: ExactLayerCoordinates =
  EXACT_LAYER_COORDINATES;

export const WANKEL_FIVE_FORCE_LIFECYCLE_TRACE: WankelFiveForceLifecycleTrace =
  ['fork', 'race', 'fold', 'vent', 'interfere'];

export const EXACT_SATURATION_TRACE: ExactSaturationTrace = [10, 30, 90];

export const BRAID_FOLD_ANCHOR_COORDINATES: BraidFoldAnchorCoordinates = [
  10, 17, 22,
];

export const EXACT_DECOMPOSITION_TRIPLE_COORDINATES: ExactDecompositionTripleCoordinates =
  [7, 17, 22];
export const EXACT_INTERFERENCE_STRIDE_TRACE: ExactInterferenceStrideTrace = [
  33, 31, 26, 21,
];
export const EXACT_PRESSURE_WEIGHTED_INTERFERENCE_STRIDE_TRACE: ExactPressureWeightedInterferenceStrideTrace =
  [0, 558, 2028];
export const EXACT_PRESSURE_WEIGHTED_INTERFERENCE_TOTAL = 2586 as const;
export const PRESSURE_CERTIFIED_EXECUTION_PATH_VERSION =
  'pressure-certified-execution-path/v1' as const;
export const PRESSURE_CERTIFIED_EXECUTION_PATH_IDENTITY: ExactPressureCertifiedExecutionPathIdentity =
  'gnosis-uring__topology-10-12-17-22-43__interfere-43-55__pressure-2586SPa__closure-30210SPa';
export const EXACT_EXTENDED_AEON_RANGES: ExactExtendedAeonRanges = [
  [1, 12],
  [10, 22],
  [20, 32],
  [30, 43],
  [50, 55],
];
export const EXACT_EXTENDED_AEON_KEYSTONE_TRACE: ExactExtendedAeonKeystoneTrace =
  [
    [7, 11],
    [17, 22],
    [27, 32],
    [37, 43],
    [55, 55],
  ];
export const GNOSIS_URING_INTERFERE_PORT_COORDINATE = 43 as const;
export const GNOSIS_URING_INTERFERE_CLOSURE_COORDINATE = 55 as const;

export const LOCAL_AEON_ROW_BUDGET = 12 as const;
export const WANKEL_AEON_CELLS = LOCAL_AEON_ROW_BUDGET;
export const WANKEL_RUNTIME_DEADLINE_PULSE_HOPS = 264 as const;
export const STATISTICAL_TOPOLOGICAL_VOLUME = LOCAL_AEON_ROW_BUDGET;
export const LOCAL_AEON_CARRYING_CAPACITY = 144 as const;
export const GNOSTIC_GRAM_UNIT = 'gG' as const;
export const SIEVE_PASCAL_UNIT = 'SPa' as const;
export const PRESSURE_GRADIENT_MULTIPLIER = 8 as const;
export const EXACT_VISIBLE_MASS_TRACE: ExactVisibleMassTrace = [10, 12, 12];
export const EXACT_LEAKAGE_TRACE: ExactLeakageTrace = [0, 18, 78];
export const EXACT_INFORMATION_MASS_TRACE: ExactInformationMassTrace = [
  120, 360, 1080,
];
export const EXACT_LEAKAGE_MASS_TRACE: ExactLeakageMassTrace = [0, 216, 936];
export const EXACT_NOISE_BANDS: readonly ['white', 'pink', 'brown'] = [
  'white',
  'pink',
  'brown',
];
export const EXACT_PRESSURE_RANKED_BANDS: readonly ['brown', 'pink', 'white'] =
  ['brown', 'pink', 'white'];
export const EXACT_SPHERICAL_ROTATION_TRACE: ExactSphericalRotationTrace = [
  120, 360, 1080,
];
export const EXACT_SPHERICAL_WRAP_TRACE: ExactSphericalWrapTrace = [1, 1, 3];
export const EXACT_PACKED_GEOMETRY_BUCKET_WEIGHT = 130 as const;
export const EXACT_LENS_OBSERVATION_COORDINATES: ExactLensObservationCoordinates =
  [12, 30, 90, 43];
export const EXACT_LENS_OBSERVATION_SHAPES: ExactLensObservationShapes = [
  'plane',
  'sphere',
  'torus',
  'singularAxis',
];
export const EXACT_PORT_LENS_OBSERVATION_COORDINATES: ExactPortLensObservationCoordinates =
  [10, 12, 17, 22, 43];
export const EXACT_PORT_LENS_OBSERVATION_SHAPES: ExactPortLensObservationShapes =
  ['plane', 'plane', 'torus', 'torus', 'singularAxis'];
export const EXACT_CENTERLINE_GAMUT_TRACE: ExactCenterlineGamutTrace = [
  10, 12, 17, 22, 43,
];
export const EXACT_CYCLIC_COMPILER_REPRESENTATIVE_TRACE: ExactCyclicCompilerRepresentativeTrace =
  [180, 180];
export const EXACT_CYCLIC_COMPILER_RECONSTRUCTED_TRACE: ExactCyclicCompilerReconstructedTrace =
  [360, 1080];
export const EXACT_CLOSED_LOOP_TRACE: ExactClosedLoopTrace = [
  10, 12, 17, 22, 43, 55, 10,
];
export const EXACT_PACKED_LENS_ROLE_TRACE: ExactPackedLensRoleTrace = [
  'coreFilament',
  'innerRing',
  'outerBraidedShell',
];
export const EXACT_PACKED_LENS_SHAPE_TRACE: ExactPackedLensShapeTrace = [
  'plane',
  'sphere',
  'torus',
];
export const PERFECT_LOCAL_DOUBLE_COVER_COORDINATE = 24 as const;
export const DOUBLE_KEYSTONE_COVER_DEFICIT = 2 as const;
export const INTERFERENCE_CLOSURE_FAMILY_INDEX = 1 as const;
export const SECOND_INTERFERENCE_CLOSURE_FAMILY_INDEX = 2 as const;
export const THIRD_INTERFERENCE_CLOSURE_FAMILY_INDEX = 3 as const;
export const MAX_BOUNDED_INTERFERENCE_CLOSURE_FAMILY_INDEX = 5 as const;
export const BOUNDED_INTERFERENCE_CLOSURE_FAMILY_INDICES = [
  1, 2, 3, 4, 5,
] as const;
export const INTERFERENCE_CLOSURE_STRIDE = 12 as const;
export const GNOSIS_URING_SECOND_INTERFERE_CLOSURE_COORDINATE = 67 as const;
export const GNOSIS_URING_THIRD_INTERFERE_CLOSURE_COORDINATE = 79 as const;
export const GNOSIS_URING_FOURTH_INTERFERE_CLOSURE_COORDINATE = 91 as const;
export const GNOSIS_URING_FIFTH_INTERFERE_CLOSURE_COORDINATE = 103 as const;
export const INTERFERENCE_CLOSURE_PRESSURE_GROWTH_BLOCK = 1152 as const;
export const GNOSIS_URING_FIRST_CLOSURE_AXIS_PRESSURE_TOTAL = 3738 as const;
export const GNOSIS_URING_SECOND_CLOSURE_AXIS_PRESSURE_TOTAL = 4890 as const;
export const GNOSIS_URING_THIRD_CLOSURE_AXIS_PRESSURE_TOTAL = 6042 as const;
export const GNOSIS_URING_FOURTH_CLOSURE_AXIS_PRESSURE_TOTAL = 7194 as const;
export const GNOSIS_URING_FIFTH_CLOSURE_AXIS_PRESSURE_TOTAL = 8346 as const;
export const GNOSIS_URING_BOUNDED_CLOSURE_FAMILY_PRESSURE_TOTAL =
  30210 as const;
export const EXACT_CLOSED_LOOP_RUNTIME_METADATA_CHECKSUM = 6439 as const;
export const EXACT_PRESSURE_CERTIFIED_EXECUTION_PATH_IDENTITY =
  PRESSURE_CERTIFIED_EXECUTION_PATH_IDENTITY;

/**
 * Handles the gnosis pressure Weighted Benchmark Topology Coordinates For workflow.
 */
export function pressureWeightedBenchmarkTopologyCoordinatesFor(
  interfereCoordinate: number
): readonly number[] {
  return [
    ...NATURAL_PORT_COORDINATES,
    Math.max(0, Math.floor(interfereCoordinate)),
  ];
}

/**
 * Handles the gnosis pressure Weighted Benchmark Interference Stride Trace For workflow.
 */
export function pressureWeightedBenchmarkInterferenceStrideTraceFor(
  interfereCoordinate: number
): readonly number[] {
  const anchor = Math.max(0, Math.floor(interfereCoordinate));
  return NATURAL_PORT_COORDINATES.map((coordinate) =>
    Math.max(0, anchor - coordinate)
  );
}

/**
 * Handles the gnosis pressure Weighted Benchmark Weighted Stride Trace For workflow.
 */
export function pressureWeightedBenchmarkWeightedStrideTraceFor(
  interfereCoordinate: number,
  leakageTrace: readonly number[] = EXACT_LEAKAGE_TRACE
): readonly number[] {
  const strideTrace =
    pressureWeightedBenchmarkInterferenceStrideTraceFor(interfereCoordinate);
  const count = Math.min(strideTrace.length, leakageTrace.length);
  return Array.from({ length: count }, (_, index) => {
    const stride = strideTrace[index] ?? 0;
    const leakagePressure = Math.max(0, Math.floor(leakageTrace[index] ?? 0));
    return stride * leakagePressure;
  });
}

/**
 * Handles the gnosis pressure Weighted Benchmark Total For workflow.
 */
export function pressureWeightedBenchmarkTotalFor(
  interfereCoordinate: number,
  leakageTrace: readonly number[] = EXACT_LEAKAGE_TRACE
): number {
  return pressureWeightedBenchmarkWeightedStrideTraceFor(
    interfereCoordinate,
    leakageTrace
  ).reduce((total, pressure) => total + pressure, 0);
}

export const BRAID_FOLD_THEOREM_IDS = [
  'GamutBraid.exact_braid_fork_race_fold_pass',
  'GamutBraid.fork_race_fold_braid_pass_returns_to_base',
] as const;

export const NATURAL_PORT_SYSTEM_THEOREM_IDS = [
  'GamutBraid.exact_port_action_trace',
  'GamutBraid.exact_natural_port_coordinates',
  'GamutBraid.natural_port_coordinates_match_braid_gates',
  'GamutBraid.exact_port_action_mapping',
  'GamutBraid.vent_hands_off_to_fork',
  'GamutBraid.iterate_port_action_four_cycle',
  'GamutBraid.infinite_port_action_four_cycle',
  'GamutBraid.infinite_port_coordinate_four_cycle',
  'GamutBraid.infinite_port_stream_periodic',
  'GamutBraid.exact_first_infinite_port_cycle',
  'GamutBraid.first_cycle_is_natural_stack',
  'GamutBraid.infinite_port_stage_has_action',
  'GamutBraid.exact_vent_to_fork_boundary',
  'GamutBraid.exact_vent_stage_hands_to_next_fork_stage',
] as const;

export const DECOMPOSITION_TRIPLE_THEOREM_IDS = [
  'GamutBraid.exact_seven_seventeen_twenty_two_decomposition_pass',
  'GamutBraid.decomposition_fork_race_fold_returns_to_base',
] as const;

export const BRAID_PACKET_THEOREM_IDS = [
  ...NATURAL_PORT_SYSTEM_THEOREM_IDS,
  'GamutBraid.exact_braid_packet_action_trace',
  'GamutBraid.exact_braid_packet_port_coordinates',
  'GamutBraid.exact_braid_packet_saturation_trace',
  'GamutBraid.exact_braid_packet_information_mass_trace',
  'GamutBraid.exact_braid_packet_leakage_trace',
  'GamutBraid.exact_braid_packet_total_information_mass',
  'GamutBraid.exact_braid_packet_total_leakage',
  'GamutBraid.exact_braid_packet_boundary_valid',
  'GamutBraid.exact_braid_packet_certified',
  'StatisticalMechanicsAccounting.exact_information_mass_trace',
  'StatisticalMechanicsAccounting.exact_leakage_trace',
  'StatisticalMechanicsAccounting.exact_total_information_mass',
  'StatisticalMechanicsAccounting.exact_total_leakage',
] as const;

export const EXTENDED_AEON_TRANSITION_THEOREM_IDS = [
  ...BRAID_PACKET_THEOREM_IDS,
  'GamutBraid.exact_extended_aeon_ranges',
  'GamutBraid.exact_extended_aeon_keystone_trace',
  'GamutBraid.exact_third_aeon_keystones_shift_by_ten',
  'GamutBraid.exact_fourth_aeon_fold_shift_by_ten',
  'GamutBraid.exact_fourth_aeon_prime_terminal_transition',
  'GamutBraid.exact_fifty_five_ouroboros_closure',
  'GamutBraid.exact_extended_aeon_stack_certified',
] as const;

export const INDEXED_INTERFERE_PORT_CLOSURE_FAMILY_THEOREM_IDS = [
  ...EXTENDED_AEON_TRANSITION_THEOREM_IDS,
  'GamutBraid.indexed_interfere_closure_coordinate_strictly_increases',
  'GamutBraid.indexed_interfere_closure_distance_from_exact_admission',
  'GamutBraid.indexed_interfere_closure_axis_pressure_total',
  'GamutBraid.exact_indexed_interfere_port_closure_family_member_certified',
] as const;

export const INDEXED_INTERFERE_PORT_FIRST_CLOSURE_FAMILY_THEOREM_IDS = [
  ...INDEXED_INTERFERE_PORT_CLOSURE_FAMILY_THEOREM_IDS,
  'GamutBraid.exact_indexed_interfere_port_closure_family_certified',
  'GamutBraid.exact_indexed_interfere_port_first_closure_axis_pressure_total',
] as const;

export const INDEXED_INTERFERE_PORT_SECOND_CLOSURE_FAMILY_THEOREM_IDS = [
  ...INDEXED_INTERFERE_PORT_CLOSURE_FAMILY_THEOREM_IDS,
  'GamutBraid.exact_indexed_interfere_port_second_closure_family_certified',
  'GamutBraid.exact_indexed_interfere_port_second_closure_family_distance',
  'GamutBraid.exact_indexed_interfere_port_second_closure_axis_pressure_total',
] as const;

export const INDEXED_INTERFERE_PORT_THIRD_CLOSURE_FAMILY_THEOREM_IDS = [
  ...INDEXED_INTERFERE_PORT_CLOSURE_FAMILY_THEOREM_IDS,
  'GamutBraid.exact_indexed_interfere_port_third_closure_family_distance',
  'GamutBraid.exact_indexed_interfere_port_third_closure_axis_pressure_total',
] as const;

export const INDEXED_INTERFERE_PORT_CLOSURE_FAMILY_AUDIT_THEOREM_IDS = [
  ...INDEXED_INTERFERE_PORT_CLOSURE_FAMILY_THEOREM_IDS,
  'GamutBraid.indexed_interfere_closure_family_audit_row_certified',
  'GamutBraid.indexed_interfere_closure_family_audit_row_pressure_total',
  'GamutBraid.indexed_interfere_closure_family_audit_row_distance',
  'GamutBraid.indexed_interfere_closure_family_audit_trace_members_certified',
  'GamutBraid.indexed_interfere_closure_family_audit_trace_pressure_total_sum',
] as const;

export const BOUNDED_INDEXED_INTERFERE_PORT_CLOSURE_FAMILY_AUDIT_THEOREM_IDS = [
  ...INDEXED_INTERFERE_PORT_CLOSURE_FAMILY_AUDIT_THEOREM_IDS,
  'GamutBraid.bounded_indexed_interfere_closure_family_audit_trace_certified',
  'GamutBraid.exact_bounded_indexed_interfere_closure_family_audit_trace',
  'GamutBraid.exact_bounded_indexed_interfere_closure_family_audit_pressure_total',
] as const;

export const INDEXED_INTERFERE_PORT_ADMISSION_THEOREM_IDS = [
  ...INDEXED_INTERFERE_PORT_FIRST_CLOSURE_FAMILY_THEOREM_IDS,
  ...BOUNDED_INDEXED_INTERFERE_PORT_CLOSURE_FAMILY_AUDIT_THEOREM_IDS,
  'GamutBraid.exact_indexed_interfere_port_admission_certified',
  'GamutBraid.indexed_interfere_port_admission_forces_exact_coordinates',
] as const;

export const INDEXED_INTERFERE_PORT_WRONG_CLOSURE_REJECTION_THEOREM_IDS = [
  ...INDEXED_INTERFERE_PORT_SECOND_CLOSURE_FAMILY_THEOREM_IDS,
  'GamutBraid.indexed_interfere_closure_distance_positive_for_later',
  'GamutBraid.exact_indexed_interfere_port_third_closure_family_distance',
  'GamutBraid.indexed_interfere_port_later_closure_family_rejected_by_exact_admission',
  'GamutBraid.indexed_interfere_port_second_closure_family_rejected_by_exact_admission',
] as const;

export const GNOSIS_URING_INTERFERE_PORT_THEOREM_IDS = [
  ...INDEXED_INTERFERE_PORT_ADMISSION_THEOREM_IDS,
  ...INDEXED_INTERFERE_PORT_WRONG_CLOSURE_REJECTION_THEOREM_IDS,
  'GamutBraid.exact_interfere_port_action_mapping',
  'GamutBraid.interfere_port_is_terminal',
  'GamutBraid.exact_interfere_port_coordinate',
  'GamutBraid.exact_interfere_port_closure_coordinate',
  'GamutBraid.exact_interfere_port_certified',
] as const;

export const INTERFERENCE_STRIDE_THEOREM_IDS = [
  ...GNOSIS_URING_INTERFERE_PORT_THEOREM_IDS,
  'GamutBraid.exact_interfere_port_external_to_natural_stack',
  'GamutBraid.exact_interference_stride_trace',
  'GamutBraid.exact_interference_anchor_packet_certified',
] as const;

export const PRESSURE_WEIGHTED_INTERFERENCE_STRIDE_THEOREM_IDS = [
  ...INTERFERENCE_STRIDE_THEOREM_IDS,
  'GamutBraid.exact_pressure_weighted_interference_stride_trace',
  'GamutBraid.exact_pressure_weighted_interference_total',
  'GamutBraid.exact_pressure_weighted_interference_anchor_packet_certified',
  'GamutBraid.exact_pressure_weighted_interfere_port_closure_family_first_member',
  'GamutBraid.pressure_weighted_interfere_port_closure_family_first_member_forces_exact_coordinates',
  'GamutBraid.exact_pressure_weighted_interfere_port_closure_family_member_certified',
  'GamutBraid.pressure_weighted_interfere_port_later_closure_family_rejected_by_exact_admission',
  'GamutBraid.exact_pressure_weighted_interfere_port_second_closure_family_certified',
] as const;

export const CYCLIC_QUOTIENT_COMPILER_THEOREM_IDS = [
  'GamutBraid.exact_pink_antipodal_cyclic_quotient_certified',
  'GamutBraid.exact_brown_triple_wrap_cyclic_quotient_certified',
  'GamutBraid.exact_brown_antipodal_cyclic_quotient_certified',
  'GamutBraid.exact_brown_sixfold_cyclic_quotient_certified',
  'GamutBraid.exact_cyclic_compiler_full_orbit_trace',
  'GamutBraid.exact_cyclic_compiler_representative_trace',
  'GamutBraid.exact_cyclic_compiler_reconstructed_trace',
  'GamutBraid.exact_cyclic_compiler_full_work',
  'GamutBraid.exact_cyclic_compiler_representative_work',
  'GamutBraid.exact_cyclic_compiler_fivefold_work_reduction',
  'GamutBraid.exact_cyclic_compiler_reduces_work',
  'GamutBraid.exact_cyclic_quotient_compiler_certified',
] as const;

export const GNOSTIC_LENS_GEOMETRY_THEOREM_IDS = [
  ...PRESSURE_WEIGHTED_INTERFERENCE_STRIDE_THEOREM_IDS,
  ...CYCLIC_QUOTIENT_COMPILER_THEOREM_IDS,
  'GamutBraid.exact_spherical_accounting_trace',
  'GamutBraid.exact_spherical_saturation_trace',
  'GamutBraid.exact_spherical_information_mass_trace',
  'GamutBraid.exact_spherical_rotation_trace',
  'GamutBraid.exact_spherical_wrap_trace',
  'GamutBraid.kenoma_to_pink_saturation_triples',
  'GamutBraid.pink_to_brown_saturation_triples',
  'GamutBraid.kenoma_to_pink_mass_triples',
  'GamutBraid.pink_to_brown_mass_triples',
  'GamutBraid.brown_rotation_is_three_pink_rotations',
  'GamutBraid.brown_mass_is_three_pink_masses',
  'GamutBraid.packed_geometry_total_mass_is_fifteen_sixty',
  'GamutBraid.packed_geometry_bucket_ratio',
  'GamutBraid.packed_geometry_hidden_aeon_weight',
  'GamutBraid.pink_antipodal_symmetry',
  'GamutBraid.brown_antipodal_symmetry',
  'GamutBraid.pink_harmonic_quotient_certified',
  'GamutBraid.brown_harmonic_quotient_certified',
  'GamutBraid.pink_harmonic_shortcut_halves_quadratic_work',
  'GamutBraid.brown_harmonic_shortcut_halves_quadratic_work',
  'GamutBraid.pink_harmonic_shortcut_reduces_work',
  'GamutBraid.brown_harmonic_shortcut_reduces_work',
  'GamutBraid.high_latitude_ports_are_prime_and_double',
  'GamutBraid.exact_spherical_accounting_certified',
  'GamutBraid.exact_lens_perspective_trace_certified',
  'GamutBraid.exact_dual_geometry_certified',
  'GamutBraid.exact_port_lens_perspective_certified',
  'GamutBraid.exact_toroidal_braid_nested_in_spherical_enclosure',
  'GamutBraid.double_keystone_cover_deficit_is_two',
  'GamutBraid.double_keystone_is_two_eleven_carriers',
  'GamutBraid.fold_swerve_between_bucket_and_double',
  'GamutBraid.exact_twist_pressure_trace',
  'GamutBraid.exact_twisted_double_keystone_certified',
  'GamutBraid.twist_pressure_exceeds_flat_reference',
  'GamutBraid.exact_toroidal_axis_certified',
  'GamutBraid.exact_interference_closure_stride_is_bucket',
  'GamutBraid.exact_centerline_gamut_trace',
  'GamutBraid.exact_packed_lens_role_trace',
  'GamutBraid.exact_packed_lens_mass_trace',
  'GamutBraid.exact_packed_lens_shape_trace',
  'GamutBraid.exact_packed_lens_shell_certified',
  'GamutBraid.exact_centerline_gamut_certified',
  'GamutBraid.exact_gnostic_lens_geometry_certified',
  'GamutBraid.exact_gnostic_lens_verdict_certified',
  'GamutBraid.exact_self_referential_knot_certified',
] as const;

export const CLOSED_LOOP_RUNTIME_METADATA_THEOREM_IDS = [
  ...GNOSTIC_LENS_GEOMETRY_THEOREM_IDS,
  'GamutBraid.exact_closed_loop_trace',
  'GamutBraid.exact_closed_loop_certified',
  'GamutBraid.exact_closed_loop_weight_accumulation_representative_shortcut',
  'GamutBraid.exact_closed_loop_repeated_unit_workload_certified',
  'GamutBraid.exact_pressure_weighted_closed_loop_workload_certified',
  'GamutBraid.exact_closed_loop_runtime_metadata_checksum',
  'GamutBraid.exact_closed_loop_runtime_metadata_certified',
] as const;

export const PRESSURE_WEIGHTED_CLOSED_LOOP_BENCHMARK_REPORT_THEOREM_IDS = [
  'GamutBraid.exact_closed_loop_weight_accumulation_representative_shortcut',
  'GamutBraid.exact_closed_loop_repeated_unit_workload_certified',
  'GamutBraid.exact_pressure_weighted_closed_loop_workload_certified',
  'GamutBraid.exact_pressure_weighted_closed_loop_benchmark_report_family_certified',
  'GamutBraid.exact_pressure_weighted_closed_loop_benchmark_report_certified',
] as const;

export const PRESSURE_CERTIFIED_EXECUTION_PATH_STABLE_TIE_ANCHOR_THEOREM_IDS = [
  'StatisticalMechanicsAccounting.exact_indexed_pressure_selector_stable_tie_contract',
  'StatisticalMechanicsAccounting.exact_indexed_pressure_selector_stable_tie_selects_first_route_id',
  'StatisticalMechanicsAccounting.exact_indexed_pressure_selector_stable_tie_runtime_order_alignment',
  'StatisticalMechanicsAccounting.exact_indexed_pressure_selector_stable_tie_preserves_input_index_zero',
  'StatisticalMechanicsAccounting.exact_indexed_pressure_selector_stable_tie_rejects_over_budget',
  'StatisticalMechanicsAccounting.exact_indexed_pressure_selector_stable_tie_rejects_route_id',
  'StatisticalMechanicsAccounting.exact_indexed_pressure_selector_stable_tie_rejects_pressure',
  'StatisticalMechanicsAccounting.exact_indexed_pressure_selector_stable_tie_admits_tight_route_under_bucket',
  'StatisticalMechanicsAccounting.exact_indexed_pressure_selector_stable_tie_lower_rejection_audit',
] as const;

export const STATISTICAL_MECHANICS_THEOREM_IDS = [
  'StatisticalMechanicsAccounting.exact_statistical_mechanics_accounting_complete',
  'StatisticalMechanicsAccounting.exact_gnostic_metrology_complete',
  'StatisticalMechanicsAccounting.exact_pressure_gradient_certificate',
  'StatisticalMechanicsAccounting.exact_calibrated_fluidic_pressure_bridge',
  'StatisticalMechanicsAccounting.exact_fluidic_routing_admissible',
  'StatisticalMechanicsAccounting.pressure_budget_tightening_lower_rejects_route',
  'StatisticalMechanicsAccounting.pressure_budget_tightening_rejects_or_preserves_order',
  'StatisticalMechanicsAccounting.exact_pressure_budget_same_order_preserved',
  'StatisticalMechanicsAccounting.exact_pressure_budget_lower_rejection_audit',
  'StatisticalMechanicsAccounting.exact_pressure_selector_picks_lowest_route',
  'StatisticalMechanicsAccounting.exact_pressure_selector_admits_lowest_route',
  'StatisticalMechanicsAccounting.exact_pressure_selector_rejects_over_budget',
  'StatisticalMechanicsAccounting.exact_pressure_selector_same_order_audit',
  'StatisticalMechanicsAccounting.exact_pressure_selector_lower_rejection_audit',
  'StatisticalMechanicsAccounting.exact_indexed_pressure_selector_picks_primary_route',
  'StatisticalMechanicsAccounting.exact_indexed_pressure_selector_admits_primary_route',
  'StatisticalMechanicsAccounting.exact_indexed_pressure_selector_selects_prime_route_id',
  'StatisticalMechanicsAccounting.exact_indexed_pressure_selector_pressure_projection',
  'StatisticalMechanicsAccounting.exact_indexed_pressure_selector_selected_candidate_has_minimum_pressure',
  'StatisticalMechanicsAccounting.exact_indexed_pressure_selector_stable_tie_contract',
  'StatisticalMechanicsAccounting.exact_indexed_pressure_selector_stable_tie_selects_first_route_id',
  'StatisticalMechanicsAccounting.exact_indexed_pressure_selector_stable_tie_runtime_order_alignment',
  'StatisticalMechanicsAccounting.exact_indexed_pressure_selector_stable_tie_preserves_input_index_zero',
  'StatisticalMechanicsAccounting.exact_indexed_pressure_selector_stable_tie_rejects_over_budget',
  'StatisticalMechanicsAccounting.exact_indexed_pressure_selector_stable_tie_rejects_route_id',
  'StatisticalMechanicsAccounting.exact_indexed_pressure_selector_stable_tie_rejects_pressure',
  'StatisticalMechanicsAccounting.exact_indexed_pressure_selector_stable_tie_admits_tight_route_under_bucket',
  'StatisticalMechanicsAccounting.exact_indexed_pressure_selector_stable_tie_lower_rejection_audit',
  'StatisticalMechanicsAccounting.exact_indexed_pressure_selector_same_order_audit',
  'StatisticalMechanicsAccounting.exact_indexed_pressure_selector_lower_rejection_audit',
  'StatisticalMechanicsAccounting.lower_indexed_route_pressure_left_biased_on_tie',
  'StatisticalMechanicsAccounting.lowest_indexed_route_candidate_from_preserves_seed_on_ties',
  'StatisticalMechanicsAccounting.lowest_indexed_route_candidate_stable_tie_contract',
  'StatisticalMechanicsAccounting.select_indexed_route_within_budget_stable_tie_contract',
  'StatisticalMechanicsAccounting.select_indexed_route_within_budget_stable_tie_runtime_order_alignment',
  'StatisticalMechanicsAccounting.select_indexed_route_within_budget_stable_tie_preserves_input_index_zero',
  'StatisticalMechanicsAccounting.select_indexed_route_within_budget_stable_tie_rejects_over_budget',
  'StatisticalMechanicsAccounting.select_indexed_route_within_budget_stable_tie_rejects_route_id',
  'StatisticalMechanicsAccounting.select_indexed_route_within_budget_stable_tie_rejects_pressure',
  'StatisticalMechanicsAccounting.lowest_indexed_route_candidate_minimum_correct',
  'StatisticalMechanicsAccounting.select_indexed_route_within_budget_minimum_correct',
  'StatisticalMechanicsAccounting.lowest_indexed_route_candidate_membership_preserved',
  'StatisticalMechanicsAccounting.select_indexed_route_within_budget_membership_preserved',
  'StatisticalMechanicsAccounting.lowest_indexed_route_candidate_certified_selector_contract',
  'StatisticalMechanicsAccounting.select_indexed_route_within_budget_certified_selector_contract',
  'StatisticalMechanicsAccounting.exact_indexed_pressure_selector_certified_primary_route_contract',
  'StatisticalMechanicsAccounting.exact_indexed_pressure_selector_certified_tight_route_contract',
  'StatisticalMechanicsAccounting.layer_handshake_determines_leakage_trace',
  'StatisticalMechanicsAccounting.exact_total_information_mass',
] as const;

export const PRESSURE_CERTIFIED_EXECUTION_PATH_THEOREM_IDS = [
  'GamutBraid.exact_pressure_certified_execution_path',
  'GamutBraid.pressure_certified_execution_path_forces_topology',
  'GamutBraid.pressure_certified_execution_path_forces_interfere_port',
  'GamutBraid.pressure_certified_execution_path_pressure_totals',
  'GamutBraid.pressure_certified_execution_path_closure_family_checksum',
  'GamutBraid.pressure_certified_execution_path_benchmark_identity',
  'GamutBraid.pressure_certified_execution_path_stable_tie_contract',
  ...GNOSTIC_LENS_GEOMETRY_THEOREM_IDS,
  ...GNOSIS_URING_INTERFERE_PORT_THEOREM_IDS,
  ...PRESSURE_WEIGHTED_INTERFERENCE_STRIDE_THEOREM_IDS,
  ...BOUNDED_INDEXED_INTERFERE_PORT_CLOSURE_FAMILY_AUDIT_THEOREM_IDS,
  ...PRESSURE_WEIGHTED_CLOSED_LOOP_BENCHMARK_REPORT_THEOREM_IDS,
  ...STATISTICAL_MECHANICS_THEOREM_IDS,
] as const;

export const WANKEL_RUNTIME_CERTIFICATE_THEOREM_IDS = [
  ...CLOSED_LOOP_RUNTIME_METADATA_THEOREM_IDS,
  ...PRESSURE_WEIGHTED_CLOSED_LOOP_BENCHMARK_REPORT_THEOREM_IDS,
  ...PRESSURE_CERTIFIED_EXECUTION_PATH_THEOREM_IDS,
  'WankelEngineTheorem.wankel_schedules_interference',
  'RuntimeCertificate.wankel_fifth_force_mechanism_has_runtime_certificate',
  'RuntimeCertificate.wankel_fast_path_preserves_runtime_result_equivalence',
  'RuntimeCertificate.wankel_fast_path_cost_dominates_baseline',
  'RuntimeCertificate.wankel_aggregate_fast_path_cost_dominates_baseline',
  'RuntimeCertificate.wankel_robust_fast_path_estimator_dominates_baseline',
] as const;

export interface ClosedArchitectureCertificate {
  readonly source: 'GnosisClosedArchitectureCertificate';
  readonly layerCoordinates: ExactLayerCoordinates;
  readonly saturationTrace: ExactSaturationTrace;
  readonly keystoneGap: 5;
  readonly baseReady: true;
  readonly closedArchitecture: true;
  readonly returnsToBase: true;
  readonly theoremIds: readonly [
    'CosmicArchitecture.exact_cosmic_architecture_closed',
    'CosmicArchitecture.snake_eats_tail_returns_to_base'
  ];
}

export interface SaturationObservation {
  readonly saturation: number;
  readonly expectedLayer: CosmosLayer | null;
  readonly focus: SieveFocus | null;
  /**
   * Lossy admit/reject collapse, preserved for the decision-tree and
   * reason-string consumers. `admitted === isAccept(verdict)`: an observation
   * that lands on a keystone gate is admitted; everything else (deferred OR
   * disqualified) collapses to `false`, exactly as before.
   */
  readonly admitted: boolean;
  /**
   * The canonical ternary lift of the lifecycle admission. A saturation that
   * lands on an exact keystone gate (10/30/90) is qualified + committed to a
   * layer → Accept. A finite saturation strictly inside the active keystone
   * band (between the first and last gate) but off any gate is qualified yet
   * contingent — it has not yet reached a scheduled layer → Abstain (defer).
   * Anything else (non-finite, negative, or outside the band) is disqualified
   * → Decline.
   */
  readonly verdict: Verdict;
}

export interface NaturalPortSystemCertificate {
  readonly source: 'GnosisNaturalPortSystemCertificate';
  readonly actionTrace: ExactPortActionTrace;
  readonly portCoordinates: ExactLayerCoordinates;
  readonly firstInfiniteCycleCoordinates: ExactLayerCoordinates;
  readonly cycleLength: 4;
  readonly ventHandsOffToFork: true;
  readonly ventToForkBoundary: NaturalPortCycleBoundary;
  readonly coordinatesMatchBraidGates: true;
  readonly returnsToBase: true;
  readonly theoremIds: typeof NATURAL_PORT_SYSTEM_THEOREM_IDS;
}

export interface NaturalPortCycleBoundary {
  readonly source: 'GnosisNaturalPortCycleBoundary';
  readonly outgoingStage: 3;
  readonly incomingStage: 4;
  readonly outgoingAction: 'vent';
  readonly incomingAction: 'fork';
  readonly outgoingCoordinate: 22;
  readonly incomingCoordinate: 10;
  readonly contiguousStages: true;
  readonly theoremIds: typeof NATURAL_PORT_SYSTEM_THEOREM_IDS;
}

export interface BraidFoldCandidate {
  readonly cycle: number;
  readonly gateCoordinates: readonly number[];
  readonly saturationTrace: readonly number[];
  readonly closedArchitecture: boolean;
  readonly endpointsLinked: boolean;
  readonly baseReady: boolean;
}

export interface DecompositionTriple {
  readonly decompositionGate: number;
  readonly raceAnchor: number;
  readonly foldAnchor: number;
}

export interface DecompositionTripleCandidate {
  readonly cycle: number;
  readonly coordinates: readonly number[];
}

export interface BraidFoldCertificate {
  readonly source: 'GnosisBraidFoldCertificate';
  readonly cycle: number;
  readonly layerCoordinates: ExactLayerCoordinates;
  readonly saturationTrace: ExactSaturationTrace;
  readonly forkedAnchorCoordinates: BraidFoldAnchorCoordinates;
  readonly primeDoubleCrossingCount: 5;
  readonly closedArchitecture: true;
  readonly endpointsLinked: true;
  readonly baseReady: true;
  readonly returnsToBase: true;
  readonly theoremIds: typeof BRAID_FOLD_THEOREM_IDS;
}

export interface DecompositionTripleCertificate {
  readonly source: 'GnosisDecompositionTripleCertificate';
  readonly cycle: number;
  readonly coordinates: ExactDecompositionTripleCoordinates;
  readonly triple: {
    readonly decompositionGate: 7;
    readonly raceAnchor: 17;
    readonly foldAnchor: 22;
  };
  readonly decompositionFeedsKenomaSpan: true;
  readonly anchorsKeystones: true;
  readonly primeDoubleCrossingCount: 5;
  readonly braidFoldCertificate: BraidFoldCertificate;
  readonly returnsToBase: true;
  readonly theoremIds: typeof DECOMPOSITION_TRIPLE_THEOREM_IDS;
}

export interface BraidPacketCertificate {
  readonly source: 'GnosisBraidPacketCertificate';
  readonly cycle: number;
  readonly actionTrace: ExactPortActionTrace;
  readonly portCoordinates: ExactLayerCoordinates;
  readonly saturationTrace: ExactSaturationTrace;
  readonly informationMassTrace: ExactInformationMassTrace;
  readonly leakageTrace: ExactLeakageTrace;
  readonly totalInformationMass: 1560;
  readonly totalLeakage: 96;
  readonly boundary: NaturalPortCycleBoundary;
  readonly accounting: StatisticalMechanicsAccountingCertificate;
  readonly returnsToBase: true;
  readonly theoremIds: typeof BRAID_PACKET_THEOREM_IDS;
}

export interface ExtendedAeonPrimeTransition {
  readonly aeon: 'fourth';
  readonly linearTerminal: 42;
  readonly clinamen: 1;
  readonly terminal: 43;
  readonly foldKeystone: 37;
  readonly ventKeystone: 43;
}

export interface ExtendedAeonTransitionCertificate {
  readonly source: 'GnosisExtendedAeonTransitionCertificate';
  readonly ranges: ExactExtendedAeonRanges;
  readonly keystoneTrace: ExactExtendedAeonKeystoneTrace;
  readonly primeTransition: ExtendedAeonPrimeTransition;
  readonly closure: {
    readonly coordinate: 55;
    readonly carrier: 11;
    readonly cycles: 5;
  };
  readonly braidPacket: BraidPacketCertificate;
  readonly returnsToBase: true;
  readonly theoremIds: typeof EXTENDED_AEON_TRANSITION_THEOREM_IDS;
}

export interface GnosisUringInterferePortCertificate {
  readonly source: 'GnosisUringInterferePortCertificate';
  readonly host: 'gnosis-uring';
  readonly portAction: 'interfere';
  readonly portCoordinate: typeof GNOSIS_URING_INTERFERE_PORT_COORDINATE;
  readonly closureCoordinate: typeof GNOSIS_URING_INTERFERE_CLOSURE_COORDINATE;
  readonly terminal: true;
  readonly indexedAdmission: IndexedInterferePortAdmissionCertificate;
  readonly closureFamilyAudit: IndexedInterferePortClosureFamilyAuditCertificate;
  readonly transition: ExtendedAeonTransitionCertificate;
  readonly returnsToBase: true;
  readonly theoremIds: typeof GNOSIS_URING_INTERFERE_PORT_THEOREM_IDS;
}

export interface IndexedInterferePortAdmissionCertificate {
  readonly source: 'GnosisIndexedInterferePortAdmissionCertificate';
  readonly portAction: 'interfere';
  readonly portCoordinate: typeof GNOSIS_URING_INTERFERE_PORT_COORDINATE;
  readonly closureCoordinate: typeof GNOSIS_URING_INTERFERE_CLOSURE_COORDINATE;
  readonly closureIndex: typeof INTERFERENCE_CLOSURE_FAMILY_INDEX;
  readonly closureStride: typeof INTERFERENCE_CLOSURE_STRIDE;
  readonly closureFamily: IndexedInterferePortClosureFamilyCertificate;
  readonly terminal: true;
  readonly transition: ExtendedAeonTransitionCertificate;
  readonly returnsToBase: true;
  readonly theoremIds: typeof INDEXED_INTERFERE_PORT_ADMISSION_THEOREM_IDS;
}

export interface IndexedInterferePortClosureFamilyCertificate {
  readonly source: 'GnosisIndexedInterferePortClosureFamilyCertificate';
  readonly portAction: 'interfere';
  readonly portCoordinate: typeof GNOSIS_URING_INTERFERE_PORT_COORDINATE;
  readonly closureCoordinate: number;
  readonly closureIndex: number;
  readonly closureStride: typeof INTERFERENCE_CLOSURE_STRIDE;
  readonly closureStepsAfterExactAdmission: number;
  readonly closureDistanceFromExactAdmission: number;
  readonly closureAxisPressureWeightedTotal: number;
  readonly closureAxisPressureGrowthFromInterferencePort: number;
  readonly closureAxisPressureGrowthFromExactClosure: number;
  readonly orderedAfterExactAdmission: boolean;
  readonly terminal: true;
  readonly exactAdmissionCompatible: boolean;
  readonly transition: ExtendedAeonTransitionCertificate;
  readonly returnsToBase: true;
  readonly theoremIds:
    | typeof INDEXED_INTERFERE_PORT_CLOSURE_FAMILY_THEOREM_IDS
    | typeof INDEXED_INTERFERE_PORT_FIRST_CLOSURE_FAMILY_THEOREM_IDS
    | typeof INDEXED_INTERFERE_PORT_SECOND_CLOSURE_FAMILY_THEOREM_IDS
    | typeof INDEXED_INTERFERE_PORT_THIRD_CLOSURE_FAMILY_THEOREM_IDS;
}

export interface IndexedInterferePortClosureFamilyAuditCertificate {
  readonly source: 'GnosisIndexedInterferePortClosureFamilyAuditCertificate';
  readonly portAction: 'interfere';
  readonly portCoordinate: typeof GNOSIS_URING_INTERFERE_PORT_COORDINATE;
  readonly closureStride: typeof INTERFERENCE_CLOSURE_STRIDE;
  readonly maxClosureIndex: typeof MAX_BOUNDED_INTERFERENCE_CLOSURE_FAMILY_INDEX;
  readonly closureIndices: readonly number[];
  readonly members: readonly IndexedInterferePortClosureFamilyCertificate[];
  readonly pressureGrowthBlock: typeof INTERFERENCE_CLOSURE_PRESSURE_GROWTH_BLOCK;
  readonly closureFamilyPressureTotal: number;
  readonly pressureUnit: typeof SIEVE_PASCAL_UNIT;
  readonly exactAdmissionClosureIndex: typeof INTERFERENCE_CLOSURE_FAMILY_INDEX;
  readonly exactAdmissionClosureCoordinate: typeof GNOSIS_URING_INTERFERE_CLOSURE_COORDINATE;
  readonly returnsToBase: true;
  readonly theoremIds:
    | typeof INDEXED_INTERFERE_PORT_CLOSURE_FAMILY_AUDIT_THEOREM_IDS
    | typeof BOUNDED_INDEXED_INTERFERE_PORT_CLOSURE_FAMILY_AUDIT_THEOREM_IDS;
}

export interface IndexedInterferePortClosureFamilyRejectionCertificate {
  readonly source: 'GnosisIndexedInterferePortClosureFamilyRejectionCertificate';
  readonly rejectedBy: 'GnosisIndexedInterferePortAdmissionCertificate';
  readonly reason: 'wrong-closure-index';
  readonly expectedClosureCoordinate: typeof GNOSIS_URING_INTERFERE_CLOSURE_COORDINATE;
  readonly expectedClosureIndex: typeof INTERFERENCE_CLOSURE_FAMILY_INDEX;
  readonly family: IndexedInterferePortClosureFamilyCertificate;
  readonly closureDistanceFromExactAdmission: number;
  readonly returnsToBase: true;
  readonly theoremIds: typeof INDEXED_INTERFERE_PORT_WRONG_CLOSURE_REJECTION_THEOREM_IDS;
}

export interface InterferenceStrideCertificate {
  readonly source: 'GnosisInterferenceStrideCertificate';
  readonly portAction: 'interfere';
  readonly anchorCoordinate: typeof GNOSIS_URING_INTERFERE_PORT_COORDINATE;
  readonly closureCoordinate: typeof GNOSIS_URING_INTERFERE_CLOSURE_COORDINATE;
  readonly naturalCoordinates: ExactLayerCoordinates;
  readonly strideTrace: ExactInterferenceStrideTrace;
  readonly externalToNaturalStack: true;
  readonly terminal: true;
  readonly braidPacket: BraidPacketCertificate;
  readonly returnsToBase: true;
  readonly theoremIds: typeof INTERFERENCE_STRIDE_THEOREM_IDS;
}

export interface PressureWeightedInterferenceStrideCertificate {
  readonly source: 'GnosisPressureWeightedInterferenceStrideCertificate';
  readonly portAction: 'interfere';
  readonly anchorCoordinate: typeof GNOSIS_URING_INTERFERE_PORT_COORDINATE;
  readonly closureCoordinate: typeof GNOSIS_URING_INTERFERE_CLOSURE_COORDINATE;
  readonly strideTrace: ExactInterferenceStrideTrace;
  readonly leakageTrace: ExactLeakageTrace;
  readonly leakageTotal: 96;
  readonly weightedPressureTrace: ExactPressureWeightedInterferenceStrideTrace;
  readonly weightedPressureTotal: typeof EXACT_PRESSURE_WEIGHTED_INTERFERENCE_TOTAL;
  readonly pressureUnit: typeof SIEVE_PASCAL_UNIT;
  readonly weighting: 'stride-leakage-zip';
  readonly interferenceStride: InterferenceStrideCertificate;
  readonly returnsToBase: true;
  readonly theoremIds: typeof PRESSURE_WEIGHTED_INTERFERENCE_STRIDE_THEOREM_IDS;
}

export interface SphericalAccountingLayerMirror {
  readonly layer: PackedGeometryLayer;
  readonly shape: PackedTopologyShape;
  readonly saturation: number;
  readonly informationMass: number;
  readonly rotationDegrees: number;
  readonly wrapCount: number;
}

export interface FiniteSphericalHarmonicShortcutMirror {
  readonly shell: 'pink' | 'brown';
  readonly fullRotationDegrees: 360 | 1080;
  readonly sampledDegrees: 180 | 540;
  readonly mirroredDegrees: 180 | 540;
  readonly quotientFactor: 2;
  readonly fullQuadraticWork: number;
  readonly quotientQuadraticWork: number;
  readonly halvesQuadraticWork: true;
  readonly reducesWork: true;
}

export interface LensObservationMirror {
  readonly coordinate: number;
  readonly shape: LensShape;
}

export interface PortLensObservationMirror extends LensObservationMirror {
  readonly role: PortLensRole;
}

export interface TwistedDoubleKeystoneCertificate {
  readonly bucket: typeof LOCAL_AEON_ROW_BUDGET;
  readonly halfCarrier: 11;
  readonly doubleKeystone: 22;
  readonly perfectDoubleCover: typeof PERFECT_LOCAL_DOUBLE_COVER_COORDINATE;
  readonly coverDeficit: typeof DOUBLE_KEYSTONE_COVER_DEFICIT;
  readonly foldSwerve: 17;
  readonly crossingCount: 5;
  readonly leakagePressure: 96;
  readonly flatReferencePressure: 0;
  readonly pressureExceedsFlatReference: true;
}

export interface GnosticLensVerdictMirror {
  readonly localResolution: typeof LOCAL_AEON_ROW_BUDGET;
  readonly executionPathShape: 'torus';
  readonly informationContainerShape: 'sphere';
  readonly terminalInterferenceShape: 'singularAxis';
  readonly totalPackedMass: 1560;
  readonly centerlineTrace: ExactCenterlineGamutTrace;
  readonly finiteSelfReferenceShape: 'selfReferenceKnot';
}

export interface GnosticLensGeometryCertificate {
  readonly source: 'GnosisGnosticLensGeometryCertificate';
  readonly sphericalAccountingTrace: readonly [
    SphericalAccountingLayerMirror,
    SphericalAccountingLayerMirror,
    SphericalAccountingLayerMirror
  ];
  readonly sphericalSaturationTrace: ExactSaturationTrace;
  readonly sphericalInformationMassTrace: ExactInformationMassTrace;
  readonly sphericalRotationTrace: ExactSphericalRotationTrace;
  readonly sphericalWrapTrace: ExactSphericalWrapTrace;
  readonly packedGeometryTotalMass: 1560;
  readonly packedGeometryBucketWeight: typeof EXACT_PACKED_GEOMETRY_BUCKET_WEIGHT;
  readonly highLatitudePortCoordinates: readonly [17, 22];
  readonly harmonicShortcuts: readonly [
    FiniteSphericalHarmonicShortcutMirror,
    FiniteSphericalHarmonicShortcutMirror
  ];
  readonly lensObservationCoordinates: ExactLensObservationCoordinates;
  readonly lensObservationShapes: ExactLensObservationShapes;
  readonly portLensObservationCoordinates: ExactPortLensObservationCoordinates;
  readonly portLensObservationShapes: ExactPortLensObservationShapes;
  readonly portLensObservations: readonly [
    PortLensObservationMirror,
    PortLensObservationMirror,
    PortLensObservationMirror,
    PortLensObservationMirror,
    PortLensObservationMirror
  ];
  readonly nestedGeometry: {
    readonly enclosureShape: 'sphere';
    readonly braidShape: 'torus';
    readonly enclosureRotationDegrees: 360;
    readonly braidRotationDegrees: 1080;
    readonly braidWrapCount: 3;
    readonly brownRotationIsThreePinkRotations: true;
  };
  readonly twistedDoubleKeystone: TwistedDoubleKeystoneCertificate;
  readonly toroidalAxis: {
    readonly coordinate: typeof GNOSIS_URING_INTERFERE_PORT_COORDINATE;
    readonly externalToNaturalStack: true;
    readonly terminal: true;
    readonly closureCoordinate: typeof GNOSIS_URING_INTERFERE_CLOSURE_COORDINATE;
    readonly closureStride: typeof INTERFERENCE_CLOSURE_STRIDE;
  };
  readonly packedLensRoleTrace: ExactPackedLensRoleTrace;
  readonly packedLensMassTrace: ExactInformationMassTrace;
  readonly packedLensShapeTrace: ExactPackedLensShapeTrace;
  readonly centerlineGamutTrace: ExactCenterlineGamutTrace;
  readonly verdict: GnosticLensVerdictMirror;
  readonly braidPacket: BraidPacketCertificate;
  readonly returnsToBase: true;
  readonly theoremIds: typeof GNOSTIC_LENS_GEOMETRY_THEOREM_IDS;
}

export interface ClosedLoopRuntimeMetadataCertificate {
  readonly source: 'GnosisClosedLoopRuntimeMetadataCertificate';
  readonly cycle: number;
  readonly closedTrace: ExactClosedLoopTrace;
  readonly centerlineTrace: ExactCenterlineGamutTrace;
  readonly representativeTrace: ExactCyclicCompilerRepresentativeTrace;
  readonly reconstructedTrace: ExactCyclicCompilerReconstructedTrace;
  readonly axisCoordinate: typeof GNOSIS_URING_INTERFERE_PORT_COORDINATE;
  readonly closureCoordinate: typeof GNOSIS_URING_INTERFERE_CLOSURE_COORDINATE;
  readonly totalMass: 1560;
  readonly totalLeakage: 96;
  readonly pressureWeightedTotal: typeof EXACT_PRESSURE_WEIGHTED_INTERFERENCE_TOTAL;
  readonly bucketRatio: typeof EXACT_PACKED_GEOMETRY_BUCKET_WEIGHT;
  readonly checksum: typeof EXACT_CLOSED_LOOP_RUNTIME_METADATA_CHECKSUM;
  readonly gnosticLens: GnosticLensGeometryCertificate;
  readonly pressureWeightedInterference: PressureWeightedInterferenceStrideCertificate;
  readonly returnsToBase: true;
  readonly theoremIds: typeof CLOSED_LOOP_RUNTIME_METADATA_THEOREM_IDS;
}

export interface PressureWeightedClosedLoopBenchmarkReportCertificate {
  readonly source: 'GnosisPressureWeightedClosedLoopBenchmarkReportCertificate';
  readonly topologyCoordinates: ExactCenterlineGamutTrace;
  readonly gateCoordinates: ExactLayerCoordinates;
  readonly decompositionCoordinates: ExactDecompositionTripleCoordinates;
  readonly primeDoubleCrossingCount: 5;
  readonly interferePortCoordinate: typeof GNOSIS_URING_INTERFERE_PORT_COORDINATE;
  readonly interfereClosureCoordinate: typeof GNOSIS_URING_INTERFERE_CLOSURE_COORDINATE;
  readonly interferenceStrideTrace: ExactInterferenceStrideTrace;
  readonly leakagePressureTrace: ExactLeakageTrace;
  readonly pressureWeightedInterferenceStrideTrace: ExactPressureWeightedInterferenceStrideTrace;
  readonly pressureWeightedInterferenceTotal: typeof EXACT_PRESSURE_WEIGHTED_INTERFERENCE_TOTAL;
  readonly pressureUnit: typeof SIEVE_PASCAL_UNIT;
  readonly representativeWorkloadTheoremIds: readonly [
    'GamutBraid.exact_closed_loop_weight_accumulation_representative_shortcut',
    'GamutBraid.exact_closed_loop_repeated_unit_workload_certified',
    'GamutBraid.exact_pressure_weighted_closed_loop_workload_certified'
  ];
  readonly closedLoopRuntimeMetadata: ClosedLoopRuntimeMetadataCertificate;
  readonly pressureWeightedInterference: PressureWeightedInterferenceStrideCertificate;
  readonly returnsToBase: true;
  readonly theoremIds: typeof PRESSURE_WEIGHTED_CLOSED_LOOP_BENCHMARK_REPORT_THEOREM_IDS;
}

export interface PressureCertifiedExecutionPathInterferePort {
  readonly action: 'interfere';
  readonly coordinate: typeof GNOSIS_URING_INTERFERE_PORT_COORDINATE;
  readonly closureCoordinate: typeof GNOSIS_URING_INTERFERE_CLOSURE_COORDINATE;
  readonly terminal: true;
}

export interface PressureCertifiedExecutionPathPressureWeightedInterference {
  readonly strideTrace: ExactInterferenceStrideTrace;
  readonly leakagePressureTrace: ExactLeakageTrace;
  readonly pressureWeightedStrideTrace: ExactPressureWeightedInterferenceStrideTrace;
  readonly pressureWeightedTotal: typeof EXACT_PRESSURE_WEIGHTED_INTERFERENCE_TOTAL;
  readonly pressureUnit: typeof SIEVE_PASCAL_UNIT;
}

export interface PressureCertifiedExecutionPathMetrology {
  readonly informationMassTotal: 1560;
  readonly leakageTotal: 96;
  readonly massUnit: typeof GNOSTIC_GRAM_UNIT;
  readonly pressureUnit: typeof SIEVE_PASCAL_UNIT;
}

export interface PressureCertifiedExecutionPathCertificate {
  readonly source: 'GnosisPressureCertifiedExecutionPathCertificate';
  readonly certificateVersion: typeof PRESSURE_CERTIFIED_EXECUTION_PATH_VERSION;
  readonly pressureIdentity: typeof PRESSURE_CERTIFIED_EXECUTION_PATH_IDENTITY;
  readonly runtime: 'gnosis';
  readonly host: 'gnosis-uring';
  readonly topologyCoordinates: ExactCenterlineGamutTrace;
  readonly gateCoordinates: ExactLayerCoordinates;
  readonly decompositionCoordinates: ExactDecompositionTripleCoordinates;
  readonly primeDoubleCrossingCount: 5;
  readonly interferePort: PressureCertifiedExecutionPathInterferePort;
  readonly pressureWeightedInterference: PressureCertifiedExecutionPathPressureWeightedInterference;
  readonly closureFamilyAudit: IndexedInterferePortClosureFamilyAuditCertificate;
  readonly informationMassTotal: 1560;
  readonly leakageTotal: 96;
  readonly metrology: PressureCertifiedExecutionPathMetrology;
  readonly pressureWeightedClosedLoopBenchmarkReport: PressureWeightedClosedLoopBenchmarkReportCertificate;
  readonly returnsToBase: true;
  readonly theoremIds: typeof PRESSURE_CERTIFIED_EXECUTION_PATH_THEOREM_IDS;
}

export interface WankelRuntimeCertificate {
  readonly source: 'GnosisWankelRuntimeCertificate';
  readonly runtime: 'gnosis';
  readonly host: 'gnosis-uring';
  readonly aeonCells: typeof WANKEL_AEON_CELLS;
  readonly lifecycleTrace: WankelFiveForceLifecycleTrace;
  readonly deadlinePulseHops: typeof WANKEL_RUNTIME_DEADLINE_PULSE_HOPS;
  readonly closedLoopRuntimeMetadata: ClosedLoopRuntimeMetadataCertificate;
  readonly pressureWeightedClosedLoopBenchmarkReport: PressureWeightedClosedLoopBenchmarkReportCertificate;
  readonly pressureCertifiedExecutionPath: PressureCertifiedExecutionPathCertificate;
  readonly returnsToBase: true;
  readonly theoremIds: typeof WANKEL_RUNTIME_CERTIFICATE_THEOREM_IDS;
}

export interface StatisticalMechanicsAccountingCertificate {
  readonly source: 'GnosisStatisticalMechanicsAccountingCertificate';
  readonly layerCoordinates: ExactLayerCoordinates;
  readonly saturationTrace: ExactSaturationTrace;
  readonly visibleRowBudget: typeof LOCAL_AEON_ROW_BUDGET;
  readonly topologicalVolume: typeof STATISTICAL_TOPOLOGICAL_VOLUME;
  readonly visibleMassTrace: ExactVisibleMassTrace;
  readonly leakageTrace: ExactLeakageTrace;
  readonly informationMassTrace: ExactInformationMassTrace;
  readonly leakageMassTrace: ExactLeakageMassTrace;
  readonly totalLeakage: 96;
  readonly totalInformationMass: 1560;
  readonly constrainedByWitnesses: true;
  readonly returnsToBase: true;
  readonly theoremIds: typeof STATISTICAL_MECHANICS_THEOREM_IDS;
}

export interface GnosticMetrologyLayerRow {
  readonly band: StatisticalNoiseBand;
  readonly saturation: number;
  readonly visibleMass: number;
  readonly sievePressure: number;
  readonly gnosticMass: number;
  readonly leakageMass: number;
}

export interface GnosticMetrologyProfile {
  readonly source: 'GnosisGnosticMetrologyProfile';
  readonly massUnit: typeof GNOSTIC_GRAM_UNIT;
  readonly pressureUnit: typeof SIEVE_PASCAL_UNIT;
  readonly localCarryingCapacity: typeof LOCAL_AEON_CARRYING_CAPACITY;
  readonly virtualMass: 1416;
  readonly totalGnosticMass: 1560;
  readonly totalSievePressure: 96;
  readonly checksumMass: 1560;
  readonly shootGamutPressureRequirement: 96;
  readonly brownBufferMass: 1080;
  readonly brownLeakageMass: 936;
  readonly layerRows: readonly [
    GnosticMetrologyLayerRow,
    GnosticMetrologyLayerRow,
    GnosticMetrologyLayerRow
  ];
  readonly returnsToBase: true;
  readonly theoremIds: typeof STATISTICAL_MECHANICS_THEOREM_IDS;
}

export interface PressureGradientCertificate {
  readonly source: 'GnosisPressureGradientCertificate';
  readonly bucketSize: typeof LOCAL_AEON_ROW_BUDGET;
  readonly bucketCount: typeof PRESSURE_GRADIENT_MULTIPLIER;
  readonly totalSievePressure: 96;
  readonly bucketProduct: 96;
  readonly reserveSievePressure: 96;
  readonly returnsToBase: true;
  readonly theoremIds: typeof STATISTICAL_MECHANICS_THEOREM_IDS;
}

export interface GnosticRuntimeAllocationPlan {
  readonly source: 'GnosisRuntimeAllocationPlan';
  readonly observedMass: number;
  readonly expectedMass: 1560;
  readonly checksumMatches: boolean;
  readonly reserveBrownBufferMass: 1080;
  readonly reserveSievePressure: 96;
  readonly pressureGradient: PressureGradientCertificate;
  readonly admitted: boolean;
  readonly reason: 'checksum-match' | 'checksum-mismatch';
  readonly theoremIds: typeof STATISTICAL_MECHANICS_THEOREM_IDS;
}

export interface CalibratedFluidicPressureBridge {
  readonly source: 'GnosisCalibratedFluidicPressureBridge';
  readonly calibrationPaPerSPa: number;
  readonly pressureTracePa: readonly [number, number, number];
  readonly totalPhysicalPressurePa: number;
  readonly pinkPhysicalPressurePa: number;
  readonly brownPhysicalPressurePa: number;
  readonly pressureGradient: PressureGradientCertificate;
  readonly admitted: true;
  readonly reason: 'calibrated-pressure-bridge';
  readonly theoremIds: typeof STATISTICAL_MECHANICS_THEOREM_IDS;
}

export interface FluidicRoutingAdmissibilityCertificate {
  readonly source: 'GnosisFluidicRoutingAdmissibilityCertificate';
  readonly calibrationPaPerSPa: number;
  readonly observedMass: number;
  readonly expectedMass: 1560;
  readonly checksumMatches: true;
  readonly pressureRankedBands: typeof EXACT_PRESSURE_RANKED_BANDS;
  readonly dominantBand: 'brown';
  readonly dominantSievePressure: 78;
  readonly dominantPhysicalPressurePa: number;
  readonly pinkPhysicalPressurePa: number;
  readonly totalPhysicalPressurePa: number;
  readonly preservesGradient: true;
  readonly allocationPlan: GnosticRuntimeAllocationPlan;
  readonly fluidicBridge: CalibratedFluidicPressureBridge;
  readonly pressureGradient: PressureGradientCertificate;
  readonly admitted: true;
  readonly returnsToBase: true;
  readonly reason: 'positive-calibration-checksum-match';
  readonly theoremIds: typeof STATISTICAL_MECHANICS_THEOREM_IDS;
}

export interface BraidFoldAudit {
  /**
   * Lossy accept/reject collapse, preserved for the reason-string consumers.
   * `accepted === isAccept(verdict)`: only a fully certified candidate is
   * accepted; both hard structural rejections and deferred-pending-readiness
   * candidates collapse to `false`, exactly as before.
   */
  readonly accepted: boolean;
  /**
   * The canonical ternary lift of the audit. A fully certified candidate →
   * Accept. A candidate of the wrong canonical shape (gate coordinates,
   * saturation trace, or crossing count that can never be the braid fold) is
   * disqualified → Decline. A candidate of the RIGHT shape whose readiness
   * flags are not yet asserted (closed architecture / linked endpoints / base
   * readiness absent) is a genuine candidate that is simply unscheduled →
   * Abstain (defer): it may be re-audited once those flags are supplied.
   */
  readonly verdict: Verdict;
  readonly reason: string;
  readonly anchorCoordinates: readonly number[];
  readonly certificate: BraidFoldCertificate | null;
}

export interface DecompositionTripleAudit {
  readonly accepted: boolean;
  readonly reason: string;
  readonly coordinates: readonly number[];
  readonly certificate: DecompositionTripleCertificate | null;
}

export interface CertifiedBraidFastPath {
  readonly source: 'GnosisCertifiedBraidFastPath';
  readonly certificate: BraidFoldCertificate;
  readonly forkedAnchorCoordinates: BraidFoldAnchorCoordinates;
  readonly collapsedBridge: {
    readonly from: 17;
    readonly to: 22;
    readonly stride: 5;
  };
  readonly returnsToBase: true;
}

export interface CertifiedBraidFastPathDecision {
  /**
   * Lossy select/reject collapse, preserved for the equivalence and
   * reason-string consumers. `selected === isAccept(verdict)`: only an adopted
   * fast path is selected; both "no certificate to select" and "candidate
   * deferred" collapse to `false`, exactly as before.
   */
  readonly selected: boolean;
  /**
   * The canonical ternary lift of the fast-path selection. An adopted fast
   * path → Accept. No certificate at all (nothing to select) → Decline. A real
   * certificate that is deliberately NOT adopted because a contingent gate is
   * unmet (does not return to base) → Abstain: deliberate non-adoption of an
   * otherwise-present candidate, not a failure.
   */
  readonly verdict: Verdict;
  readonly reason: string;
  readonly path: CertifiedBraidFastPath | null;
}

export interface CertifiedBraidFastPathEquivalence<T> {
  readonly decision: CertifiedBraidFastPathDecision;
  readonly equivalent: boolean;
  readonly linearResult: T;
  readonly certifiedResult: T;
}

export interface WankelRuntimeFastPathDecision {
  /**
   * Lossy select/reject collapse, preserved for the equivalence and
   * reason-string consumers. `selected === isAccept(verdict)`: only an adopted
   * fast path is selected; "no certificate", "inexact certificate", and
   * "deferred candidate" all collapse to `false`, exactly as before.
   */
  readonly selected: boolean;
  /**
   * The canonical ternary lift of the fast-path selection. An adopted fast
   * path → Accept. No certificate (nothing to select) or an inexact / malformed
   * certificate (the wrong shape, can never be the Wankel fast path) → Decline.
   * A well-formed certificate deliberately NOT adopted because a contingent
   * gate is unmet (does not return to base) → Abstain: deliberate non-adoption
   * of an otherwise-present candidate, not a failure.
   */
  readonly verdict: Verdict;
  readonly reason: string;
  readonly certificate: WankelRuntimeCertificate | null;
  readonly deadlinePulseHops: typeof WANKEL_RUNTIME_DEADLINE_PULSE_HOPS | null;
  readonly lifecycleTrace: WankelFiveForceLifecycleTrace | null;
}

export interface WankelRuntimeFastPathEquivalence<T> {
  readonly decision: WankelRuntimeFastPathDecision;
  readonly equivalent: boolean;
  readonly linearResult: T;
  readonly certifiedResult: T;
}

export interface WankelRuntimeFastPathCostWitness {
  readonly selectedCost: number;
  readonly baselineCost: number;
  readonly unit: 'nsPerIter' | 'costUnits';
}

export interface WankelRuntimeFastPathCostDominance<T> {
  readonly equivalence: WankelRuntimeFastPathEquivalence<T>;
  readonly witness: WankelRuntimeFastPathCostWitness;
  readonly dominates: boolean;
  readonly reason: string;
  readonly theoremId:
    'RuntimeCertificate.wankel_fast_path_cost_dominates_baseline';
}

export interface WankelRuntimeFastPathAggregateCostWitness {
  readonly sampleCount: number;
  readonly selectedTotalCost: number;
  readonly baselineTotalCost: number;
  readonly unit: 'nsPerIter' | 'costUnits';
}

export interface WankelRuntimeFastPathAggregateCostDominance<T> {
  readonly equivalence: WankelRuntimeFastPathEquivalence<T>;
  readonly witness: WankelRuntimeFastPathAggregateCostWitness;
  readonly dominates: boolean;
  readonly reason: string;
  readonly theoremId:
    'RuntimeCertificate.wankel_aggregate_fast_path_cost_dominates_baseline';
}

export interface WankelRuntimeFastPathRobustEstimatorWitness {
  readonly sampleCount: number;
  readonly minimumSampleCount: number;
  readonly selectedTotalCost: number;
  readonly baselineTotalCost: number;
  readonly selectedRobustCost: number;
  readonly baselineRobustCost: number;
  readonly observedJitter: number;
  readonly jitterBudget: number;
  readonly outlierBudget: number;
  readonly unit: 'nsPerIter' | 'costUnits';
}

export interface WankelRuntimeFastPathRobustEstimatorDominance<T> {
  readonly equivalence: WankelRuntimeFastPathEquivalence<T>;
  readonly witness: WankelRuntimeFastPathRobustEstimatorWitness;
  readonly dominates: boolean;
  readonly reason: string;
  readonly theoremId:
    'RuntimeCertificate.wankel_robust_fast_path_estimator_dominates_baseline';
}

function isNonNegativeInteger(value: number): boolean {
  return Number.isInteger(value) && value >= 0;
}

function sameNumberTrace(
  left: readonly number[],
  right: readonly number[]
): boolean {
  return (
    left.length === right.length &&
    left.every((value, index) => value === right[index])
  );
}

function sameStringTrace(
  left: readonly string[],
  right: readonly string[]
): boolean {
  return (
    left.length === right.length &&
    left.every((value, index) => value === right[index])
  );
}

function anchorCoordinatesFromTrace(
  gateCoordinates: readonly number[]
): readonly number[] {
  return BRAID_FOLD_ANCHOR_COORDINATES.filter((coordinate) =>
    gateCoordinates.includes(coordinate)
  );
}

/**
 * Creates the Exact Decomposition Triple.
 */
export function createExactDecompositionTriple(): DecompositionTriple {
  return {
    decompositionGate: 7,
    raceAnchor: 17,
    foldAnchor: 22,
  };
}

/**
 * Handles the gnosis decomposition Triple Coordinates workflow.
 */
export function decompositionTripleCoordinates(
  triple: DecompositionTriple
): readonly number[] {
  return [triple.decompositionGate, triple.raceAnchor, triple.foldAnchor];
}

/**
 * Handles the gnosis decomposition Feeds Kenoma Span workflow.
 */
export function decompositionFeedsKenomaSpan(
  triple: DecompositionTriple
): boolean {
  return (
    triple.raceAnchor === triple.decompositionGate + layerCoordinate('kenoma')
  );
}

/**
 * Handles the gnosis decomposition Anchors Keystones workflow.
 */
export function decompositionAnchorsKeystones(
  triple: DecompositionTriple,
  certificate: BraidFoldCertificate | null
): boolean {
  return (
    certificate !== null &&
    triple.raceAnchor === certificate.layerCoordinates[2] &&
    triple.foldAnchor === certificate.layerCoordinates[3]
  );
}

/** Mirror of `CosmicArchitecture.layerCoordinate`. */
export function layerCoordinate(layer: CosmosLayer): number {
  switch (layer) {
    case 'kenoma':
      return 10;
    case 'localAeon':
      return 12;
    case 'primeKeystone':
      return 17;
    case 'doubleKeystone':
      return 22;
  }
}

/** Mirror of `SovereignSieve.expectedLayerForSaturation`. */
export function expectedLayerForSaturation(
  saturation: number
): CosmosLayer | null {
  switch (saturation) {
    case 10:
      return 'kenoma';
    case 30:
      return 'primeKeystone';
    case 90:
      return 'doubleKeystone';
    default:
      return null;
  }
}

/** Mirror of `SovereignSieve.focusForLayer`. */
export function focusForLayer(layer: CosmosLayer): SieveFocus {
  switch (layer) {
    case 'kenoma':
      return 'rawSubstrate';
    case 'localAeon':
      return 'observerResolution';
    case 'primeKeystone':
      return 'witnessInversion';
    case 'doubleKeystone':
      return 'hostHandshake';
  }
}

/** Mirror of `SovereignSieve.focusForSaturation`. */
export function focusForSaturation(saturation: number): SieveFocus | null {
  const layer = expectedLayerForSaturation(saturation);
  return layer === null ? null : focusForLayer(layer);
}

/**
 * Handles the gnosis observe Saturation workflow.
 */
export function observeSaturation(saturation: number): SaturationObservation {
  const expectedLayer = expectedLayerForSaturation(saturation);
  return {
    saturation,
    expectedLayer,
    focus: expectedLayer === null ? null : focusForLayer(expectedLayer),
    admitted: expectedLayer !== null,
    verdict: saturationAdmissionVerdict(saturation, expectedLayer),
  };
}

/**
 * The ternary lift of the saturation lifecycle admission. On-gate → Accept;
 * finite and strictly inside the keystone band but off-gate → Abstain (defer);
 * otherwise → Decline. The keystone band is `[first gate, last gate]` of the
 * exact saturation trace, so the neutral middle is a real, reachable third
 * state (e.g. saturation 12 sits between gates 10 and 30 → Abstain), not an
 * invented one. `isAccept(verdict)` reproduces the legacy `admitted` boolean.
 */
function saturationAdmissionVerdict(
  saturation: number,
  expectedLayer: CosmosLayer | null
): Verdict {
  if (expectedLayer !== null) {
    return Verdict.Accept;
  }
  const firstGate = EXACT_SATURATION_TRACE[0];
  const lastGate = EXACT_SATURATION_TRACE[EXACT_SATURATION_TRACE.length - 1];
  if (
    Number.isFinite(saturation) &&
    saturation > firstGate &&
    lastGate !== undefined &&
    saturation < lastGate
  ) {
    return Verdict.Abstain;
  }
  return Verdict.Decline;
}

/** Mirror of `StatisticalMechanicsAccounting.visibleMass`. */
export function visibleMass(
  saturation: number,
  rowBudget = LOCAL_AEON_ROW_BUDGET
): number {
  return Math.min(saturation, rowBudget);
}

/** Mirror of `StatisticalMechanicsAccounting.leakage`. */
export function leakage(
  saturation: number,
  rowBudget = LOCAL_AEON_ROW_BUDGET
): number {
  return Math.max(0, saturation - rowBudget);
}

/** Mirror of `StatisticalMechanicsAccounting.informationMass`. */
export function informationMass(
  saturation: number,
  topologicalVolume = STATISTICAL_TOPOLOGICAL_VOLUME
): number {
  return saturation * topologicalVolume;
}

/** Mirror of `StatisticalMechanicsAccounting.leakageMass`. */
export function leakageMass(
  saturation: number,
  topologicalVolume = STATISTICAL_TOPOLOGICAL_VOLUME
): number {
  return leakage(saturation) * topologicalVolume;
}

/**
 * Handles the gnosis visible Mass Trace workflow.
 */
export function visibleMassTrace(
  saturationTrace: readonly number[]
): readonly number[] {
  return saturationTrace.map((saturation) => visibleMass(saturation));
}

/**
 * Handles the gnosis leakage Trace workflow.
 */
export function leakageTrace(
  saturationTrace: readonly number[]
): readonly number[] {
  return saturationTrace.map((saturation) => leakage(saturation));
}

/**
 * Handles the gnosis information Mass Trace workflow.
 */
export function informationMassTrace(
  saturationTrace: readonly number[]
): readonly number[] {
  return saturationTrace.map((saturation) => informationMass(saturation));
}

/**
 * Handles the gnosis leakage Mass Trace workflow.
 */
export function leakageMassTrace(
  saturationTrace: readonly number[]
): readonly number[] {
  return saturationTrace.map((saturation) => leakageMass(saturation));
}

/**
 * Handles the gnosis trace Total workflow.
 */
export function traceTotal(trace: readonly number[]): number {
  return trace.reduce((sum, value) => sum + value, 0);
}

/** Mirror of `CosmicArchitecture.keystoneGap` over explicit coordinates. */
export function keystoneGap(
  primeKeystone: number,
  doubleKeystone: number
): number {
  return Math.max(0, doubleKeystone - primeKeystone);
}

/**
 * Handles the gnosis port Coordinate For Action workflow.
 */
export function portCoordinateForAction(action: BraidPortAction): number {
  switch (action) {
    case 'fork':
      return 10;
    case 'race':
      return 12;
    case 'fold':
      return 17;
    case 'vent':
      return 22;
    case 'interfere':
      return GNOSIS_URING_INTERFERE_PORT_COORDINATE;
  }
}

/**
 * Handles the gnosis action For Port Coordinate workflow.
 */
export function actionForPortCoordinate(
  coordinate: number
): BraidPortAction | null {
  switch (coordinate) {
    case 10:
      return 'fork';
    case 12:
      return 'race';
    case 17:
      return 'fold';
    case 22:
      return 'vent';
    case GNOSIS_URING_INTERFERE_PORT_COORDINATE:
      return 'interfere';
    default:
      return null;
  }
}

/**
 * Handles the gnosis lens Shape For Port Coordinate workflow.
 */
export function lensShapeForPortCoordinate(
  coordinate: number
): LensShape | null {
  switch (coordinate) {
    case 10:
    case 12:
      return 'plane';
    case 17:
    case 22:
      return 'torus';
    case GNOSIS_URING_INTERFERE_PORT_COORDINATE:
      return 'singularAxis';
    default:
      return null;
  }
}

/**
 * Handles the gnosis next Port Action workflow.
 */
export function nextPortAction(action: BraidPortAction): BraidPortAction {
  switch (action) {
    case 'fork':
      return 'race';
    case 'race':
      return 'fold';
    case 'fold':
      return 'vent';
    case 'vent':
      return 'fork';
    case 'interfere':
      return 'interfere';
  }
}

function assertNaturalStage(stage: number): number {
  if (!Number.isInteger(stage) || stage < 0) {
    throw new RangeError(
      `Fork/race/fold/vent stage must be a non-negative integer, got ${stage}`
    );
  }
  return stage;
}

/**
 * Handles the gnosis infinite Fork Race Fold Vent Action workflow.
 */
export function infiniteForkRaceFoldVentAction(stage: number): BraidPortAction {
  switch (assertNaturalStage(stage) % 4) {
    case 0:
      return 'fork';
    case 1:
      return 'race';
    case 2:
      return 'fold';
    default:
      return 'vent';
  }
}

/**
 * Handles the gnosis infinite Fork Race Fold Vent Coordinate workflow.
 */
export function infiniteForkRaceFoldVentCoordinate(stage: number): number {
  return portCoordinateForAction(infiniteForkRaceFoldVentAction(stage));
}

/**
 * Handles the gnosis first Infinite Port Cycle Coordinates workflow.
 */
export function firstInfinitePortCycleCoordinates(): ExactLayerCoordinates {
  return [
    infiniteForkRaceFoldVentCoordinate(0),
    infiniteForkRaceFoldVentCoordinate(1),
    infiniteForkRaceFoldVentCoordinate(2),
    infiniteForkRaceFoldVentCoordinate(3),
  ] as ExactLayerCoordinates;
}

/**
 * Handles the gnosis port Stream Is Periodic workflow.
 */
export function portStreamIsPeriodic(stage: number): boolean {
  const naturalStage = assertNaturalStage(stage);
  return (
    infiniteForkRaceFoldVentCoordinate(naturalStage + 4) ===
    infiniteForkRaceFoldVentCoordinate(naturalStage)
  );
}

/**
 * Creates the Vent To Fork Boundary.
 */
export function createVentToForkBoundary(): NaturalPortCycleBoundary {
  return {
    source: 'GnosisNaturalPortCycleBoundary',
    outgoingStage: 3,
    incomingStage: 4,
    outgoingAction: 'vent',
    incomingAction: 'fork',
    outgoingCoordinate: 22,
    incomingCoordinate: 10,
    contiguousStages: true,
    theoremIds: NATURAL_PORT_SYSTEM_THEOREM_IDS,
  };
}

/**
 * Creates the Natural Port System Certificate.
 */
export function createNaturalPortSystemCertificate(): NaturalPortSystemCertificate {
  return {
    source: 'GnosisNaturalPortSystemCertificate',
    actionTrace: EXACT_PORT_ACTION_TRACE,
    portCoordinates: NATURAL_PORT_COORDINATES,
    firstInfiniteCycleCoordinates: firstInfinitePortCycleCoordinates(),
    cycleLength: 4,
    ventHandsOffToFork: true,
    ventToForkBoundary: createVentToForkBoundary(),
    coordinatesMatchBraidGates: true,
    returnsToBase: true,
    theoremIds: NATURAL_PORT_SYSTEM_THEOREM_IDS,
  };
}

/**
 * Creates the Closed Architecture Certificate.
 */
export function createClosedArchitectureCertificate(): ClosedArchitectureCertificate {
  return {
    source: 'GnosisClosedArchitectureCertificate',
    layerCoordinates: EXACT_LAYER_COORDINATES,
    saturationTrace: EXACT_SATURATION_TRACE,
    keystoneGap: 5,
    baseReady: true,
    closedArchitecture: true,
    returnsToBase: true,
    theoremIds: [
      'CosmicArchitecture.exact_cosmic_architecture_closed',
      'CosmicArchitecture.snake_eats_tail_returns_to_base',
    ],
  };
}

/**
 * Handles the gnosis returns To Base workflow.
 */
export function returnsToBase(
  certificate:
    | ClosedArchitectureCertificate
    | NaturalPortSystemCertificate
    | BraidFoldCertificate
    | DecompositionTripleCertificate
    | BraidPacketCertificate
    | ExtendedAeonTransitionCertificate
    | IndexedInterferePortClosureFamilyCertificate
    | IndexedInterferePortClosureFamilyAuditCertificate
    | IndexedInterferePortClosureFamilyRejectionCertificate
    | IndexedInterferePortAdmissionCertificate
    | GnosisUringInterferePortCertificate
    | InterferenceStrideCertificate
    | PressureWeightedInterferenceStrideCertificate
    | GnosticLensGeometryCertificate
    | ClosedLoopRuntimeMetadataCertificate
    | PressureWeightedClosedLoopBenchmarkReportCertificate
    | PressureCertifiedExecutionPathCertificate
    | WankelRuntimeCertificate
    | StatisticalMechanicsAccountingCertificate
    | GnosticMetrologyProfile
    | PressureGradientCertificate
    | FluidicRoutingAdmissibilityCertificate
): boolean {
  return certificate.returnsToBase;
}

function layerCoordinatesFromAccountingSource(
  certificate:
    | ClosedArchitectureCertificate
    | BraidFoldCertificate
    | DecompositionTripleCertificate
): readonly number[] {
  if (certificate.source === 'GnosisDecompositionTripleCertificate') {
    return certificate.braidFoldCertificate.layerCoordinates;
  }
  return certificate.layerCoordinates;
}

function saturationTraceFromAccountingSource(
  certificate:
    | ClosedArchitectureCertificate
    | BraidFoldCertificate
    | DecompositionTripleCertificate
): readonly number[] {
  if (certificate.source === 'GnosisDecompositionTripleCertificate') {
    return certificate.braidFoldCertificate.saturationTrace;
  }
  return certificate.saturationTrace;
}

/**
 * Creates the Statistical Mechanics Accounting.
 */
export function createStatisticalMechanicsAccounting(
  certificate:
    | ClosedArchitectureCertificate
    | BraidFoldCertificate
    | DecompositionTripleCertificate
    | null = createClosedArchitectureCertificate()
): StatisticalMechanicsAccountingCertificate | null {
  if (certificate === null || !returnsToBase(certificate)) {
    return null;
  }

  const layerCoordinates = layerCoordinatesFromAccountingSource(certificate);
  const saturationTrace = saturationTraceFromAccountingSource(certificate);
  if (
    !sameNumberTrace(layerCoordinates, EXACT_LAYER_COORDINATES) ||
    !sameNumberTrace(saturationTrace, EXACT_SATURATION_TRACE)
  ) {
    return null;
  }

  const visibleTrace = visibleMassTrace(saturationTrace);
  const leakagePressureTrace = leakageTrace(saturationTrace);
  const massTrace = informationMassTrace(saturationTrace);
  const leakageDensityTrace = leakageMassTrace(saturationTrace);

  if (
    !sameNumberTrace(visibleTrace, EXACT_VISIBLE_MASS_TRACE) ||
    !sameNumberTrace(leakagePressureTrace, EXACT_LEAKAGE_TRACE) ||
    !sameNumberTrace(massTrace, EXACT_INFORMATION_MASS_TRACE) ||
    !sameNumberTrace(leakageDensityTrace, EXACT_LEAKAGE_MASS_TRACE)
  ) {
    return null;
  }

  return {
    source: 'GnosisStatisticalMechanicsAccountingCertificate',
    layerCoordinates: EXACT_LAYER_COORDINATES,
    saturationTrace: EXACT_SATURATION_TRACE,
    visibleRowBudget: LOCAL_AEON_ROW_BUDGET,
    topologicalVolume: STATISTICAL_TOPOLOGICAL_VOLUME,
    visibleMassTrace: EXACT_VISIBLE_MASS_TRACE,
    leakageTrace: EXACT_LEAKAGE_TRACE,
    informationMassTrace: EXACT_INFORMATION_MASS_TRACE,
    leakageMassTrace: EXACT_LEAKAGE_MASS_TRACE,
    totalLeakage: 96,
    totalInformationMass: 1560,
    constrainedByWitnesses: true,
    returnsToBase: true,
    theoremIds: STATISTICAL_MECHANICS_THEOREM_IDS,
  };
}

/**
 * Creates the Braid Packet Certificate.
 */
export function createBraidPacketCertificate(
  cycle: number,
  accounting: StatisticalMechanicsAccountingCertificate | null = createStatisticalMechanicsAccounting()
): BraidPacketCertificate | null {
  if (!isNonNegativeInteger(cycle)) {
    return null;
  }
  if (accounting === null || !returnsToBase(accounting)) {
    return null;
  }

  const boundary = createVentToForkBoundary();
  const boundaryMatchesStream =
    boundary.outgoingStage + 1 === boundary.incomingStage &&
    infiniteForkRaceFoldVentAction(boundary.outgoingStage) ===
      boundary.outgoingAction &&
    infiniteForkRaceFoldVentAction(boundary.incomingStage) ===
      boundary.incomingAction &&
    infiniteForkRaceFoldVentCoordinate(boundary.outgoingStage) ===
      boundary.outgoingCoordinate &&
    infiniteForkRaceFoldVentCoordinate(boundary.incomingStage) ===
      boundary.incomingCoordinate &&
    nextPortAction(boundary.outgoingAction) === boundary.incomingAction;
  if (!boundaryMatchesStream) {
    return null;
  }

  if (
    !sameNumberTrace(accounting.layerCoordinates, NATURAL_PORT_COORDINATES) ||
    !sameNumberTrace(accounting.saturationTrace, EXACT_SATURATION_TRACE) ||
    !sameNumberTrace(
      accounting.informationMassTrace,
      EXACT_INFORMATION_MASS_TRACE
    ) ||
    !sameNumberTrace(accounting.leakageTrace, EXACT_LEAKAGE_TRACE) ||
    accounting.totalInformationMass !== 1560 ||
    accounting.totalLeakage !== 96
  ) {
    return null;
  }

  return {
    source: 'GnosisBraidPacketCertificate',
    cycle,
    actionTrace: EXACT_PORT_ACTION_TRACE,
    portCoordinates: NATURAL_PORT_COORDINATES,
    saturationTrace: EXACT_SATURATION_TRACE,
    informationMassTrace: EXACT_INFORMATION_MASS_TRACE,
    leakageTrace: EXACT_LEAKAGE_TRACE,
    totalInformationMass: 1560,
    totalLeakage: 96,
    boundary,
    accounting,
    returnsToBase: true,
    theoremIds: BRAID_PACKET_THEOREM_IDS,
  };
}

/**
 * Creates the Extended Aeon Transition Certificate.
 */
export function createExtendedAeonTransitionCertificate(
  braidPacket: BraidPacketCertificate | null = createBraidPacketCertificate(0)
): ExtendedAeonTransitionCertificate | null {
  if (braidPacket === null || !returnsToBase(braidPacket)) {
    return null;
  }
  const primeTransition: ExtendedAeonPrimeTransition = {
    aeon: 'fourth',
    linearTerminal: 42,
    clinamen: 1,
    terminal: 43,
    foldKeystone: 37,
    ventKeystone: 43,
  };
  const transitionValid =
    primeTransition.linearTerminal + primeTransition.clinamen ===
      primeTransition.terminal &&
    primeTransition.terminal === 43 &&
    primeTransition.foldKeystone === 37 &&
    primeTransition.ventKeystone === 43;
  const closure = {
    coordinate: 55,
    carrier: 11,
    cycles: 5,
  } as const;
  const closureValid = closure.coordinate === closure.carrier * closure.cycles;
  if (
    !transitionValid ||
    !closureValid ||
    braidPacket.totalInformationMass !== 1560 ||
    braidPacket.totalLeakage !== 96
  ) {
    return null;
  }

  return {
    source: 'GnosisExtendedAeonTransitionCertificate',
    ranges: EXACT_EXTENDED_AEON_RANGES,
    keystoneTrace: EXACT_EXTENDED_AEON_KEYSTONE_TRACE,
    primeTransition,
    closure,
    braidPacket,
    returnsToBase: true,
    theoremIds: EXTENDED_AEON_TRANSITION_THEOREM_IDS,
  };
}

/**
 * Creates the Indexed Interfere Port Admission Certificate.
 */
export function createIndexedInterferePortAdmissionCertificate(
  transition: ExtendedAeonTransitionCertificate | null = createExtendedAeonTransitionCertificate(),
  portCoordinate: number = GNOSIS_URING_INTERFERE_PORT_COORDINATE,
  closureCoordinate: number = GNOSIS_URING_INTERFERE_CLOSURE_COORDINATE,
  closureIndex: number = INTERFERENCE_CLOSURE_FAMILY_INDEX
): IndexedInterferePortAdmissionCertificate | null {
  const closureFamily = createIndexedInterferePortClosureFamilyCertificate(
    transition,
    portCoordinate,
    closureIndex
  );
  if (closureFamily === null) {
    return null;
  }

  if (
    !Number.isInteger(closureCoordinate) ||
    closureCoordinate !== closureFamily.closureCoordinate ||
    closureCoordinate !== GNOSIS_URING_INTERFERE_CLOSURE_COORDINATE ||
    closureFamily.closureIndex !== INTERFERENCE_CLOSURE_FAMILY_INDEX ||
    !closureFamily.exactAdmissionCompatible
  ) {
    return null;
  }

  return {
    source: 'GnosisIndexedInterferePortAdmissionCertificate',
    portAction: 'interfere',
    portCoordinate: GNOSIS_URING_INTERFERE_PORT_COORDINATE,
    closureCoordinate: GNOSIS_URING_INTERFERE_CLOSURE_COORDINATE,
    closureIndex: INTERFERENCE_CLOSURE_FAMILY_INDEX,
    closureStride: INTERFERENCE_CLOSURE_STRIDE,
    closureFamily,
    terminal: true,
    transition: closureFamily.transition,
    returnsToBase: true,
    theoremIds: INDEXED_INTERFERE_PORT_ADMISSION_THEOREM_IDS,
  };
}

function indexedInterfereClosureCoordinateFor(
  portCoordinate: number,
  closureIndex: number
): number {
  return portCoordinate + closureIndex * INTERFERENCE_CLOSURE_STRIDE;
}

function closureFamilyPressureTotalFor(
  members: readonly IndexedInterferePortClosureFamilyCertificate[]
): number {
  return traceTotal(
    members.map((member) => member.closureAxisPressureWeightedTotal)
  );
}

/**
 * Creates the Indexed Interfere Port Closure Family Certificate.
 */
export function createIndexedInterferePortClosureFamilyCertificate(
  transition: ExtendedAeonTransitionCertificate | null = createExtendedAeonTransitionCertificate(),
  portCoordinate: number = GNOSIS_URING_INTERFERE_PORT_COORDINATE,
  closureIndex: number = INTERFERENCE_CLOSURE_FAMILY_INDEX
): IndexedInterferePortClosureFamilyCertificate | null {
  if (transition === null || !returnsToBase(transition)) {
    return null;
  }
  if (
    !Number.isInteger(portCoordinate) ||
    !Number.isInteger(closureIndex) ||
    portCoordinate !== GNOSIS_URING_INTERFERE_PORT_COORDINATE ||
    transition.primeTransition.terminal !== portCoordinate ||
    transition.primeTransition.ventKeystone !== portCoordinate ||
    nextPortAction('interfere') !== 'interfere' ||
    closureIndex < INTERFERENCE_CLOSURE_FAMILY_INDEX ||
    closureIndex > MAX_BOUNDED_INTERFERENCE_CLOSURE_FAMILY_INDEX
  ) {
    return null;
  }

  const exactAdmissionCompatible =
    closureIndex === INTERFERENCE_CLOSURE_FAMILY_INDEX;
  const closureStepsAfterExactAdmission = Math.max(
    0,
    closureIndex - INTERFERENCE_CLOSURE_FAMILY_INDEX
  );
  const closureDistanceFromExactAdmission =
    closureStepsAfterExactAdmission * INTERFERENCE_CLOSURE_STRIDE;
  const closureCoordinate = indexedInterfereClosureCoordinateFor(
    GNOSIS_URING_INTERFERE_PORT_COORDINATE,
    closureIndex
  );
  const closureAxisPressureWeightedTotal =
    pressureWeightedBenchmarkTotalFor(closureCoordinate);
  const closureAxisPressureGrowthFromInterferencePort =
    closureIndex * INTERFERENCE_CLOSURE_PRESSURE_GROWTH_BLOCK;
  const closureAxisPressureGrowthFromExactClosure =
    closureStepsAfterExactAdmission *
    INTERFERENCE_CLOSURE_PRESSURE_GROWTH_BLOCK;
  const theoremIds = exactAdmissionCompatible
    ? INDEXED_INTERFERE_PORT_FIRST_CLOSURE_FAMILY_THEOREM_IDS
    : closureIndex === SECOND_INTERFERENCE_CLOSURE_FAMILY_INDEX
    ? INDEXED_INTERFERE_PORT_SECOND_CLOSURE_FAMILY_THEOREM_IDS
    : closureIndex === THIRD_INTERFERENCE_CLOSURE_FAMILY_INDEX
    ? INDEXED_INTERFERE_PORT_THIRD_CLOSURE_FAMILY_THEOREM_IDS
    : INDEXED_INTERFERE_PORT_CLOSURE_FAMILY_THEOREM_IDS;
  return {
    source: 'GnosisIndexedInterferePortClosureFamilyCertificate',
    portAction: 'interfere',
    portCoordinate: GNOSIS_URING_INTERFERE_PORT_COORDINATE,
    closureCoordinate,
    closureIndex,
    closureStride: INTERFERENCE_CLOSURE_STRIDE,
    closureStepsAfterExactAdmission,
    closureDistanceFromExactAdmission,
    closureAxisPressureWeightedTotal,
    closureAxisPressureGrowthFromInterferencePort,
    closureAxisPressureGrowthFromExactClosure,
    orderedAfterExactAdmission:
      closureIndex > INTERFERENCE_CLOSURE_FAMILY_INDEX,
    terminal: true,
    exactAdmissionCompatible,
    transition,
    returnsToBase: true,
    theoremIds,
  };
}

/**
 * Creates the Indexed Interfere Port Closure Family Audit Certificate.
 */
export function createIndexedInterferePortClosureFamilyAuditCertificate(
  transition: ExtendedAeonTransitionCertificate | null = createExtendedAeonTransitionCertificate(),
  closureIndices: readonly number[] = BOUNDED_INTERFERENCE_CLOSURE_FAMILY_INDICES
): IndexedInterferePortClosureFamilyAuditCertificate | null {
  if (transition === null || !returnsToBase(transition)) {
    return null;
  }
  if (
    closureIndices.length === 0 ||
    closureIndices.some(
      (closureIndex) =>
        !Number.isInteger(closureIndex) ||
        closureIndex < INTERFERENCE_CLOSURE_FAMILY_INDEX ||
        closureIndex > MAX_BOUNDED_INTERFERENCE_CLOSURE_FAMILY_INDEX
    )
  ) {
    return null;
  }

  const members = closureIndices.map((closureIndex) =>
    createIndexedInterferePortClosureFamilyCertificate(
      transition,
      GNOSIS_URING_INTERFERE_PORT_COORDINATE,
      closureIndex
    )
  );
  if (members.some((member) => member === null)) {
    return null;
  }
  const certifiedMembers =
    members as IndexedInterferePortClosureFamilyCertificate[];
  const formulasHold = certifiedMembers.every((member) => {
    const expectedCoordinate = indexedInterfereClosureCoordinateFor(
      GNOSIS_URING_INTERFERE_PORT_COORDINATE,
      member.closureIndex
    );
    const stepsAfterExact = Math.max(
      0,
      member.closureIndex - INTERFERENCE_CLOSURE_FAMILY_INDEX
    );
    return (
      member.closureCoordinate === expectedCoordinate &&
      member.closureDistanceFromExactAdmission ===
        stepsAfterExact * INTERFERENCE_CLOSURE_STRIDE &&
      member.closureAxisPressureWeightedTotal ===
        EXACT_PRESSURE_WEIGHTED_INTERFERENCE_TOTAL +
          member.closureIndex * INTERFERENCE_CLOSURE_PRESSURE_GROWTH_BLOCK &&
      member.closureAxisPressureGrowthFromInterferencePort ===
        member.closureIndex * INTERFERENCE_CLOSURE_PRESSURE_GROWTH_BLOCK &&
      member.closureAxisPressureGrowthFromExactClosure ===
        stepsAfterExact * INTERFERENCE_CLOSURE_PRESSURE_GROWTH_BLOCK &&
      member.exactAdmissionCompatible ===
        (member.closureIndex === INTERFERENCE_CLOSURE_FAMILY_INDEX)
    );
  });
  if (!formulasHold) {
    return null;
  }
  const closureFamilyPressureTotal =
    closureFamilyPressureTotalFor(certifiedMembers);

  const theoremIds = sameNumberTrace(
    closureIndices,
    BOUNDED_INTERFERENCE_CLOSURE_FAMILY_INDICES
  )
    ? BOUNDED_INDEXED_INTERFERE_PORT_CLOSURE_FAMILY_AUDIT_THEOREM_IDS
    : INDEXED_INTERFERE_PORT_CLOSURE_FAMILY_AUDIT_THEOREM_IDS;

  return {
    source: 'GnosisIndexedInterferePortClosureFamilyAuditCertificate',
    portAction: 'interfere',
    portCoordinate: GNOSIS_URING_INTERFERE_PORT_COORDINATE,
    closureStride: INTERFERENCE_CLOSURE_STRIDE,
    maxClosureIndex: MAX_BOUNDED_INTERFERENCE_CLOSURE_FAMILY_INDEX,
    closureIndices: [...closureIndices],
    members: certifiedMembers,
    pressureGrowthBlock: INTERFERENCE_CLOSURE_PRESSURE_GROWTH_BLOCK,
    closureFamilyPressureTotal,
    pressureUnit: SIEVE_PASCAL_UNIT,
    exactAdmissionClosureIndex: INTERFERENCE_CLOSURE_FAMILY_INDEX,
    exactAdmissionClosureCoordinate: GNOSIS_URING_INTERFERE_CLOSURE_COORDINATE,
    returnsToBase: true,
    theoremIds,
  };
}

/**
 * Creates the Indexed Interfere Port Closure Family Rejection Certificate.
 */
export function createIndexedInterferePortClosureFamilyRejectionCertificate(
  family: IndexedInterferePortClosureFamilyCertificate | null = createIndexedInterferePortClosureFamilyCertificate(
    createExtendedAeonTransitionCertificate(),
    GNOSIS_URING_INTERFERE_PORT_COORDINATE,
    SECOND_INTERFERENCE_CLOSURE_FAMILY_INDEX
  )
): IndexedInterferePortClosureFamilyRejectionCertificate | null {
  if (
    family === null ||
    !returnsToBase(family) ||
    family.exactAdmissionCompatible ||
    family.closureIndex <= INTERFERENCE_CLOSURE_FAMILY_INDEX ||
    family.closureIndex > MAX_BOUNDED_INTERFERENCE_CLOSURE_FAMILY_INDEX ||
    family.closureCoordinate !==
      indexedInterfereClosureCoordinateFor(
        GNOSIS_URING_INTERFERE_PORT_COORDINATE,
        family.closureIndex
      )
  ) {
    return null;
  }

  return {
    source: 'GnosisIndexedInterferePortClosureFamilyRejectionCertificate',
    rejectedBy: 'GnosisIndexedInterferePortAdmissionCertificate',
    reason: 'wrong-closure-index',
    expectedClosureCoordinate: GNOSIS_URING_INTERFERE_CLOSURE_COORDINATE,
    expectedClosureIndex: INTERFERENCE_CLOSURE_FAMILY_INDEX,
    family,
    closureDistanceFromExactAdmission: family.closureDistanceFromExactAdmission,
    returnsToBase: true,
    theoremIds: INDEXED_INTERFERE_PORT_WRONG_CLOSURE_REJECTION_THEOREM_IDS,
  };
}

/**
 * Creates the Gnosis Uring Interfere Port Certificate.
 */
export function createGnosisUringInterferePortCertificate(
  transition: ExtendedAeonTransitionCertificate | null = createExtendedAeonTransitionCertificate()
): GnosisUringInterferePortCertificate | null {
  if (transition === null || !returnsToBase(transition)) {
    return null;
  }
  const indexedAdmission =
    createIndexedInterferePortAdmissionCertificate(transition);
  const closureFamilyAudit =
    createIndexedInterferePortClosureFamilyAuditCertificate(transition);
  if (
    indexedAdmission === null ||
    closureFamilyAudit === null ||
    portCoordinateForAction('interfere') !==
      GNOSIS_URING_INTERFERE_PORT_COORDINATE ||
    actionForPortCoordinate(GNOSIS_URING_INTERFERE_PORT_COORDINATE) !==
      'interfere' ||
    nextPortAction('interfere') !== 'interfere' ||
    transition.primeTransition.terminal !==
      GNOSIS_URING_INTERFERE_PORT_COORDINATE ||
    transition.primeTransition.ventKeystone !==
      GNOSIS_URING_INTERFERE_PORT_COORDINATE ||
    transition.closure.coordinate !== GNOSIS_URING_INTERFERE_CLOSURE_COORDINATE
  ) {
    return null;
  }

  return {
    source: 'GnosisUringInterferePortCertificate',
    host: 'gnosis-uring',
    portAction: 'interfere',
    portCoordinate: GNOSIS_URING_INTERFERE_PORT_COORDINATE,
    closureCoordinate: GNOSIS_URING_INTERFERE_CLOSURE_COORDINATE,
    terminal: true,
    indexedAdmission,
    closureFamilyAudit,
    transition,
    returnsToBase: true,
    theoremIds: GNOSIS_URING_INTERFERE_PORT_THEOREM_IDS,
  };
}

/**
 * Creates the Interference Stride Certificate.
 */
export function createInterferenceStrideCertificate(
  braidPacket: BraidPacketCertificate | null = createBraidPacketCertificate(0)
): InterferenceStrideCertificate | null {
  if (braidPacket === null || !returnsToBase(braidPacket)) {
    return null;
  }
  const strideTrace = NATURAL_PORT_COORDINATES.map(
    (coordinate) => GNOSIS_URING_INTERFERE_PORT_COORDINATE - coordinate
  );
  if (
    !sameNumberTrace(braidPacket.portCoordinates, NATURAL_PORT_COORDINATES) ||
    !sameNumberTrace(strideTrace, EXACT_INTERFERENCE_STRIDE_TRACE) ||
    braidPacket.totalInformationMass !== 1560 ||
    braidPacket.totalLeakage !== 96 ||
    NATURAL_PORT_COORDINATES.some(
      (coordinate: number) =>
        coordinate === GNOSIS_URING_INTERFERE_PORT_COORDINATE
    ) ||
    actionForPortCoordinate(GNOSIS_URING_INTERFERE_PORT_COORDINATE) !==
      'interfere' ||
    nextPortAction('interfere') !== 'interfere'
  ) {
    return null;
  }

  return {
    source: 'GnosisInterferenceStrideCertificate',
    portAction: 'interfere',
    anchorCoordinate: GNOSIS_URING_INTERFERE_PORT_COORDINATE,
    closureCoordinate: GNOSIS_URING_INTERFERE_CLOSURE_COORDINATE,
    naturalCoordinates: NATURAL_PORT_COORDINATES,
    strideTrace: EXACT_INTERFERENCE_STRIDE_TRACE,
    externalToNaturalStack: true,
    terminal: true,
    braidPacket,
    returnsToBase: true,
    theoremIds: INTERFERENCE_STRIDE_THEOREM_IDS,
  };
}

/**
 * Creates the Pressure Weighted Interference Stride Certificate.
 */
export function createPressureWeightedInterferenceStrideCertificate(
  interferenceStride: InterferenceStrideCertificate | null = createInterferenceStrideCertificate()
): PressureWeightedInterferenceStrideCertificate | null {
  if (interferenceStride === null || !returnsToBase(interferenceStride)) {
    return null;
  }
  const leakageTrace = interferenceStride.braidPacket.leakageTrace;
  const leakageTotal = traceTotal(leakageTrace);
  const weightedPressureTrace: readonly [number, number, number] = [
    interferenceStride.strideTrace[0] * leakageTrace[0],
    interferenceStride.strideTrace[1] * leakageTrace[1],
    interferenceStride.strideTrace[2] * leakageTrace[2],
  ];
  const weightedPressureTotal = traceTotal(weightedPressureTrace);
  if (
    !sameNumberTrace(
      interferenceStride.strideTrace,
      EXACT_INTERFERENCE_STRIDE_TRACE
    ) ||
    !sameNumberTrace(leakageTrace, EXACT_LEAKAGE_TRACE) ||
    leakageTotal !== 96 ||
    !sameNumberTrace(
      weightedPressureTrace,
      EXACT_PRESSURE_WEIGHTED_INTERFERENCE_STRIDE_TRACE
    ) ||
    weightedPressureTotal !== EXACT_PRESSURE_WEIGHTED_INTERFERENCE_TOTAL
  ) {
    return null;
  }

  return {
    source: 'GnosisPressureWeightedInterferenceStrideCertificate',
    portAction: 'interfere',
    anchorCoordinate: GNOSIS_URING_INTERFERE_PORT_COORDINATE,
    closureCoordinate: GNOSIS_URING_INTERFERE_CLOSURE_COORDINATE,
    strideTrace: EXACT_INTERFERENCE_STRIDE_TRACE,
    leakageTrace: EXACT_LEAKAGE_TRACE,
    leakageTotal: 96,
    weightedPressureTrace: EXACT_PRESSURE_WEIGHTED_INTERFERENCE_STRIDE_TRACE,
    weightedPressureTotal: EXACT_PRESSURE_WEIGHTED_INTERFERENCE_TOTAL,
    pressureUnit: SIEVE_PASCAL_UNIT,
    weighting: 'stride-leakage-zip',
    interferenceStride,
    returnsToBase: true,
    theoremIds: PRESSURE_WEIGHTED_INTERFERENCE_STRIDE_THEOREM_IDS,
  };
}

function fullQuadraticRotationWork(degrees: number): number {
  return degrees * degrees;
}

function quotientQuadraticRotationWork(
  sampledDegrees: number,
  fullRotationDegrees: number
): number {
  return sampledDegrees * fullRotationDegrees;
}

/**
 * Creates the Gnostic Lens Geometry Certificate.
 */
export function createGnosticLensGeometryCertificate(
  braidPacket: BraidPacketCertificate | null = createBraidPacketCertificate(0)
): GnosticLensGeometryCertificate | null {
  if (braidPacket === null || !returnsToBase(braidPacket)) {
    return null;
  }

  const pinkFullQuadraticWork = fullQuadraticRotationWork(360);
  const pinkQuotientQuadraticWork = quotientQuadraticRotationWork(180, 360);
  const pinkHalvesQuadraticWork =
    pinkQuotientQuadraticWork * 2 === pinkFullQuadraticWork;
  const pinkReducesWork = pinkQuotientQuadraticWork < pinkFullQuadraticWork;
  const brownFullQuadraticWork = fullQuadraticRotationWork(1080);
  const brownQuotientQuadraticWork = quotientQuadraticRotationWork(540, 1080);
  const brownHalvesQuadraticWork =
    brownQuotientQuadraticWork * 2 === brownFullQuadraticWork;
  const brownReducesWork = brownQuotientQuadraticWork < brownFullQuadraticWork;
  const pinkShortcut: FiniteSphericalHarmonicShortcutMirror = {
    shell: 'pink',
    fullRotationDegrees: 360,
    sampledDegrees: 180,
    mirroredDegrees: 180,
    quotientFactor: 2,
    fullQuadraticWork: pinkFullQuadraticWork,
    quotientQuadraticWork: pinkQuotientQuadraticWork,
    halvesQuadraticWork: true,
    reducesWork: true,
  };
  const brownShortcut: FiniteSphericalHarmonicShortcutMirror = {
    shell: 'brown',
    fullRotationDegrees: 1080,
    sampledDegrees: 540,
    mirroredDegrees: 540,
    quotientFactor: 2,
    fullQuadraticWork: brownFullQuadraticWork,
    quotientQuadraticWork: brownQuotientQuadraticWork,
    halvesQuadraticWork: true,
    reducesWork: true,
  };
  const portLensObservations: readonly [
    PortLensObservationMirror,
    PortLensObservationMirror,
    PortLensObservationMirror,
    PortLensObservationMirror,
    PortLensObservationMirror
  ] = [
    { role: 'forkCore', coordinate: 10, shape: 'plane' },
    { role: 'racePlane', coordinate: 12, shape: 'plane' },
    { role: 'foldTwist', coordinate: 17, shape: 'torus' },
    { role: 'ventExit', coordinate: 22, shape: 'torus' },
    { role: 'interferenceAxis', coordinate: 43, shape: 'singularAxis' },
  ];
  const portLensObservationCoordinates = portLensObservations.map(
    (observation) => observation.coordinate
  );
  const portLensObservationShapes = portLensObservations.map(
    (observation) => observation.shape
  );
  const closureStride =
    GNOSIS_URING_INTERFERE_CLOSURE_COORDINATE -
    GNOSIS_URING_INTERFERE_PORT_COORDINATE;

  if (
    !sameNumberTrace(braidPacket.portCoordinates, NATURAL_PORT_COORDINATES) ||
    !sameNumberTrace(braidPacket.saturationTrace, EXACT_SATURATION_TRACE) ||
    !sameNumberTrace(
      braidPacket.informationMassTrace,
      EXACT_INFORMATION_MASS_TRACE
    ) ||
    !sameNumberTrace(braidPacket.leakageTrace, EXACT_LEAKAGE_TRACE) ||
    braidPacket.totalInformationMass !== 1560 ||
    braidPacket.totalLeakage !== 96 ||
    traceTotal(EXACT_INFORMATION_MASS_TRACE) !== 1560 ||
    LOCAL_AEON_ROW_BUDGET !== 12 ||
    1560 / LOCAL_AEON_ROW_BUDGET !== EXACT_PACKED_GEOMETRY_BUCKET_WEIGHT ||
    EXACT_SPHERICAL_ROTATION_TRACE[1] !== 360 ||
    EXACT_SPHERICAL_ROTATION_TRACE[2] !==
      3 * EXACT_SPHERICAL_ROTATION_TRACE[1] ||
    !sameNumberTrace(EXACT_LENS_OBSERVATION_COORDINATES, [12, 30, 90, 43]) ||
    !sameStringTrace(EXACT_LENS_OBSERVATION_SHAPES, [
      'plane',
      'sphere',
      'torus',
      'singularAxis',
    ]) ||
    !sameNumberTrace(
      portLensObservationCoordinates,
      EXACT_PORT_LENS_OBSERVATION_COORDINATES
    ) ||
    !sameStringTrace(
      portLensObservationShapes,
      EXACT_PORT_LENS_OBSERVATION_SHAPES
    ) ||
    lensShapeForPortCoordinate(10) !== 'plane' ||
    lensShapeForPortCoordinate(12) !== 'plane' ||
    lensShapeForPortCoordinate(17) !== 'torus' ||
    lensShapeForPortCoordinate(22) !== 'torus' ||
    lensShapeForPortCoordinate(43) !== 'singularAxis' ||
    PERFECT_LOCAL_DOUBLE_COVER_COORDINATE !== 2 * LOCAL_AEON_ROW_BUDGET ||
    DOUBLE_KEYSTONE_COVER_DEFICIT !==
      PERFECT_LOCAL_DOUBLE_COVER_COORDINATE - 22 ||
    closureStride !== INTERFERENCE_CLOSURE_STRIDE ||
    closureStride !== LOCAL_AEON_ROW_BUDGET ||
    !pinkHalvesQuadraticWork ||
    !brownHalvesQuadraticWork ||
    !pinkReducesWork ||
    !brownReducesWork
  ) {
    return null;
  }

  return {
    source: 'GnosisGnosticLensGeometryCertificate',
    sphericalAccountingTrace: [
      {
        layer: 'kenomaCore',
        shape: 'trianglePlane',
        saturation: 10,
        informationMass: 120,
        rotationDegrees: 120,
        wrapCount: 1,
      },
      {
        layer: 'pinkShell',
        shape: 'informationSphere',
        saturation: 30,
        informationMass: 360,
        rotationDegrees: 360,
        wrapCount: 1,
      },
      {
        layer: 'brownAura',
        shape: 'toroidalBraid',
        saturation: 90,
        informationMass: 1080,
        rotationDegrees: 1080,
        wrapCount: 3,
      },
    ],
    sphericalSaturationTrace: EXACT_SATURATION_TRACE,
    sphericalInformationMassTrace: EXACT_INFORMATION_MASS_TRACE,
    sphericalRotationTrace: EXACT_SPHERICAL_ROTATION_TRACE,
    sphericalWrapTrace: EXACT_SPHERICAL_WRAP_TRACE,
    packedGeometryTotalMass: 1560,
    packedGeometryBucketWeight: EXACT_PACKED_GEOMETRY_BUCKET_WEIGHT,
    highLatitudePortCoordinates: [17, 22],
    harmonicShortcuts: [pinkShortcut, brownShortcut],
    lensObservationCoordinates: EXACT_LENS_OBSERVATION_COORDINATES,
    lensObservationShapes: EXACT_LENS_OBSERVATION_SHAPES,
    portLensObservationCoordinates: EXACT_PORT_LENS_OBSERVATION_COORDINATES,
    portLensObservationShapes: EXACT_PORT_LENS_OBSERVATION_SHAPES,
    portLensObservations,
    nestedGeometry: {
      enclosureShape: 'sphere',
      braidShape: 'torus',
      enclosureRotationDegrees: 360,
      braidRotationDegrees: 1080,
      braidWrapCount: 3,
      brownRotationIsThreePinkRotations: true,
    },
    twistedDoubleKeystone: {
      bucket: LOCAL_AEON_ROW_BUDGET,
      halfCarrier: 11,
      doubleKeystone: 22,
      perfectDoubleCover: PERFECT_LOCAL_DOUBLE_COVER_COORDINATE,
      coverDeficit: DOUBLE_KEYSTONE_COVER_DEFICIT,
      foldSwerve: 17,
      crossingCount: 5,
      leakagePressure: 96,
      flatReferencePressure: 0,
      pressureExceedsFlatReference: true,
    },
    toroidalAxis: {
      coordinate: GNOSIS_URING_INTERFERE_PORT_COORDINATE,
      externalToNaturalStack: true,
      terminal: true,
      closureCoordinate: GNOSIS_URING_INTERFERE_CLOSURE_COORDINATE,
      closureStride: INTERFERENCE_CLOSURE_STRIDE,
    },
    packedLensRoleTrace: EXACT_PACKED_LENS_ROLE_TRACE,
    packedLensMassTrace: EXACT_INFORMATION_MASS_TRACE,
    packedLensShapeTrace: EXACT_PACKED_LENS_SHAPE_TRACE,
    centerlineGamutTrace: EXACT_CENTERLINE_GAMUT_TRACE,
    verdict: {
      localResolution: LOCAL_AEON_ROW_BUDGET,
      executionPathShape: 'torus',
      informationContainerShape: 'sphere',
      terminalInterferenceShape: 'singularAxis',
      totalPackedMass: 1560,
      centerlineTrace: EXACT_CENTERLINE_GAMUT_TRACE,
      finiteSelfReferenceShape: 'selfReferenceKnot',
    },
    braidPacket,
    returnsToBase: true,
    theoremIds: GNOSTIC_LENS_GEOMETRY_THEOREM_IDS,
  };
}

/**
 * Handles the gnosis closed Loop Runtime Metadata Checksum workflow.
 */
export function closedLoopRuntimeMetadataChecksum(metadata: {
  readonly closedTrace: readonly number[];
  readonly representativeTrace: readonly number[];
  readonly reconstructedTrace: readonly number[];
  readonly axisCoordinate: number;
  readonly closureCoordinate: number;
  readonly totalMass: number;
  readonly totalLeakage: number;
  readonly pressureWeightedTotal: number;
  readonly bucketRatio: number;
}): number {
  return (
    traceTotal(metadata.closedTrace) +
    traceTotal(metadata.representativeTrace) +
    traceTotal(metadata.reconstructedTrace) +
    metadata.axisCoordinate +
    metadata.closureCoordinate +
    metadata.totalMass +
    metadata.totalLeakage +
    metadata.pressureWeightedTotal +
    metadata.bucketRatio
  );
}

/**
 * Creates the Closed Loop Runtime Metadata Certificate.
 */
export function createClosedLoopRuntimeMetadataCertificate(
  gnosticLens: GnosticLensGeometryCertificate | null = createGnosticLensGeometryCertificate(),
  pressureWeightedInterference: PressureWeightedInterferenceStrideCertificate | null = createPressureWeightedInterferenceStrideCertificate(
    gnosticLens === null
      ? null
      : createInterferenceStrideCertificate(gnosticLens.braidPacket)
  )
): ClosedLoopRuntimeMetadataCertificate | null {
  if (
    gnosticLens === null ||
    pressureWeightedInterference === null ||
    !returnsToBase(gnosticLens) ||
    !returnsToBase(pressureWeightedInterference)
  ) {
    return null;
  }

  const braidPacket = gnosticLens.braidPacket;
  const weightedBraidPacket =
    pressureWeightedInterference.interferenceStride.braidPacket;
  const checksumInput = {
    closedTrace: EXACT_CLOSED_LOOP_TRACE,
    representativeTrace: EXACT_CYCLIC_COMPILER_REPRESENTATIVE_TRACE,
    reconstructedTrace: EXACT_CYCLIC_COMPILER_RECONSTRUCTED_TRACE,
    axisCoordinate: GNOSIS_URING_INTERFERE_PORT_COORDINATE,
    closureCoordinate: GNOSIS_URING_INTERFERE_CLOSURE_COORDINATE,
    totalMass: braidPacket.totalInformationMass,
    totalLeakage: braidPacket.totalLeakage,
    pressureWeightedTotal: pressureWeightedInterference.weightedPressureTotal,
    bucketRatio: gnosticLens.packedGeometryBucketWeight,
  };
  const checksum = closedLoopRuntimeMetadataChecksum(checksumInput);

  if (
    braidPacket.cycle !== weightedBraidPacket.cycle ||
    !sameNumberTrace(
      gnosticLens.centerlineGamutTrace,
      EXACT_CENTERLINE_GAMUT_TRACE
    ) ||
    !sameNumberTrace(
      pressureWeightedInterference.interferenceStride.naturalCoordinates,
      NATURAL_PORT_COORDINATES
    ) ||
    pressureWeightedInterference.anchorCoordinate !==
      GNOSIS_URING_INTERFERE_PORT_COORDINATE ||
    pressureWeightedInterference.closureCoordinate !==
      GNOSIS_URING_INTERFERE_CLOSURE_COORDINATE ||
    gnosticLens.toroidalAxis.coordinate !==
      GNOSIS_URING_INTERFERE_PORT_COORDINATE ||
    gnosticLens.toroidalAxis.closureCoordinate !==
      GNOSIS_URING_INTERFERE_CLOSURE_COORDINATE ||
    gnosticLens.toroidalAxis.closureStride !== INTERFERENCE_CLOSURE_STRIDE ||
    braidPacket.totalInformationMass !== 1560 ||
    braidPacket.totalLeakage !== 96 ||
    pressureWeightedInterference.weightedPressureTotal !==
      EXACT_PRESSURE_WEIGHTED_INTERFERENCE_TOTAL ||
    gnosticLens.packedGeometryBucketWeight !==
      EXACT_PACKED_GEOMETRY_BUCKET_WEIGHT ||
    checksum !== EXACT_CLOSED_LOOP_RUNTIME_METADATA_CHECKSUM
  ) {
    return null;
  }

  return {
    source: 'GnosisClosedLoopRuntimeMetadataCertificate',
    cycle: braidPacket.cycle,
    closedTrace: EXACT_CLOSED_LOOP_TRACE,
    centerlineTrace: EXACT_CENTERLINE_GAMUT_TRACE,
    representativeTrace: EXACT_CYCLIC_COMPILER_REPRESENTATIVE_TRACE,
    reconstructedTrace: EXACT_CYCLIC_COMPILER_RECONSTRUCTED_TRACE,
    axisCoordinate: GNOSIS_URING_INTERFERE_PORT_COORDINATE,
    closureCoordinate: GNOSIS_URING_INTERFERE_CLOSURE_COORDINATE,
    totalMass: 1560,
    totalLeakage: 96,
    pressureWeightedTotal: EXACT_PRESSURE_WEIGHTED_INTERFERENCE_TOTAL,
    bucketRatio: EXACT_PACKED_GEOMETRY_BUCKET_WEIGHT,
    checksum: EXACT_CLOSED_LOOP_RUNTIME_METADATA_CHECKSUM,
    gnosticLens,
    pressureWeightedInterference,
    returnsToBase: true,
    theoremIds: CLOSED_LOOP_RUNTIME_METADATA_THEOREM_IDS,
  };
}

/**
 * Creates the Pressure Weighted Closed Loop Benchmark Report Certificate.
 */
export function createPressureWeightedClosedLoopBenchmarkReportCertificate(
  closedLoopRuntimeMetadata: ClosedLoopRuntimeMetadataCertificate | null = createClosedLoopRuntimeMetadataCertificate()
): PressureWeightedClosedLoopBenchmarkReportCertificate | null {
  if (
    closedLoopRuntimeMetadata === null ||
    !returnsToBase(closedLoopRuntimeMetadata)
  ) {
    return null;
  }

  const pressureWeightedInterference =
    closedLoopRuntimeMetadata.pressureWeightedInterference;
  if (
    !returnsToBase(pressureWeightedInterference) ||
    !sameNumberTrace(EXACT_CENTERLINE_GAMUT_TRACE, [10, 12, 17, 22, 43]) ||
    !sameNumberTrace(NATURAL_PORT_COORDINATES, [10, 12, 17, 22]) ||
    !sameNumberTrace(EXACT_DECOMPOSITION_TRIPLE_COORDINATES, [7, 17, 22]) ||
    !sameNumberTrace(
      pressureWeightedInterference.strideTrace,
      EXACT_INTERFERENCE_STRIDE_TRACE
    ) ||
    !sameNumberTrace(
      pressureWeightedInterference.leakageTrace,
      EXACT_LEAKAGE_TRACE
    ) ||
    !sameNumberTrace(
      pressureWeightedInterference.weightedPressureTrace,
      EXACT_PRESSURE_WEIGHTED_INTERFERENCE_STRIDE_TRACE
    ) ||
    pressureWeightedInterference.weightedPressureTotal !==
      EXACT_PRESSURE_WEIGHTED_INTERFERENCE_TOTAL ||
    pressureWeightedInterference.pressureUnit !== SIEVE_PASCAL_UNIT ||
    closedLoopRuntimeMetadata.pressureWeightedTotal !==
      EXACT_PRESSURE_WEIGHTED_INTERFERENCE_TOTAL
  ) {
    return null;
  }

  return {
    source: 'GnosisPressureWeightedClosedLoopBenchmarkReportCertificate',
    topologyCoordinates: EXACT_CENTERLINE_GAMUT_TRACE,
    gateCoordinates: NATURAL_PORT_COORDINATES,
    decompositionCoordinates: EXACT_DECOMPOSITION_TRIPLE_COORDINATES,
    primeDoubleCrossingCount: 5,
    interferePortCoordinate: GNOSIS_URING_INTERFERE_PORT_COORDINATE,
    interfereClosureCoordinate: GNOSIS_URING_INTERFERE_CLOSURE_COORDINATE,
    interferenceStrideTrace: EXACT_INTERFERENCE_STRIDE_TRACE,
    leakagePressureTrace: EXACT_LEAKAGE_TRACE,
    pressureWeightedInterferenceStrideTrace:
      EXACT_PRESSURE_WEIGHTED_INTERFERENCE_STRIDE_TRACE,
    pressureWeightedInterferenceTotal:
      EXACT_PRESSURE_WEIGHTED_INTERFERENCE_TOTAL,
    pressureUnit: SIEVE_PASCAL_UNIT,
    representativeWorkloadTheoremIds: [
      'GamutBraid.exact_closed_loop_weight_accumulation_representative_shortcut',
      'GamutBraid.exact_closed_loop_repeated_unit_workload_certified',
      'GamutBraid.exact_pressure_weighted_closed_loop_workload_certified',
    ],
    closedLoopRuntimeMetadata,
    pressureWeightedInterference,
    returnsToBase: true,
    theoremIds: PRESSURE_WEIGHTED_CLOSED_LOOP_BENCHMARK_REPORT_THEOREM_IDS,
  };
}

export interface CreatePressureCertifiedExecutionPathCertificateInput {
  readonly interferePort?: GnosisUringInterferePortCertificate | null;
  readonly closureFamilyAudit?: IndexedInterferePortClosureFamilyAuditCertificate | null;
  readonly pressureWeightedInterference?: PressureWeightedInterferenceStrideCertificate | null;
  readonly metrology?: GnosticMetrologyProfile | null;
  readonly pressureWeightedClosedLoopBenchmarkReport?: PressureWeightedClosedLoopBenchmarkReportCertificate | null;
}

/**
 * Creates the Pressure Certified Execution Path Certificate.
 */
export function createPressureCertifiedExecutionPathCertificate({
  interferePort = createGnosisUringInterferePortCertificate(),
  closureFamilyAudit,
  pressureWeightedInterference = createPressureWeightedInterferenceStrideCertificate(),
  metrology = createGnosticMetrologyProfile(),
  pressureWeightedClosedLoopBenchmarkReport = createPressureWeightedClosedLoopBenchmarkReportCertificate(),
}: CreatePressureCertifiedExecutionPathCertificateInput = {}): PressureCertifiedExecutionPathCertificate | null {
  const resolvedClosureFamilyAudit =
    closureFamilyAudit ?? interferePort?.closureFamilyAudit ?? null;

  if (
    interferePort === null ||
    resolvedClosureFamilyAudit === null ||
    pressureWeightedInterference === null ||
    metrology === null ||
    pressureWeightedClosedLoopBenchmarkReport === null ||
    !returnsToBase(interferePort) ||
    !returnsToBase(resolvedClosureFamilyAudit) ||
    !returnsToBase(pressureWeightedInterference) ||
    !returnsToBase(metrology) ||
    !returnsToBase(pressureWeightedClosedLoopBenchmarkReport)
  ) {
    return null;
  }

  const pressureWeighted =
    pressureWeightedClosedLoopBenchmarkReport.pressureWeightedInterference;
  if (
    !sameNumberTrace(
      pressureWeightedClosedLoopBenchmarkReport.topologyCoordinates,
      EXACT_CENTERLINE_GAMUT_TRACE
    ) ||
    !sameNumberTrace(
      pressureWeightedClosedLoopBenchmarkReport.gateCoordinates,
      NATURAL_PORT_COORDINATES
    ) ||
    !sameNumberTrace(
      pressureWeightedClosedLoopBenchmarkReport.decompositionCoordinates,
      EXACT_DECOMPOSITION_TRIPLE_COORDINATES
    ) ||
    pressureWeightedClosedLoopBenchmarkReport.primeDoubleCrossingCount !== 5 ||
    pressureWeightedClosedLoopBenchmarkReport.interferePortCoordinate !==
      GNOSIS_URING_INTERFERE_PORT_COORDINATE ||
    pressureWeightedClosedLoopBenchmarkReport.interfereClosureCoordinate !==
      GNOSIS_URING_INTERFERE_CLOSURE_COORDINATE ||
    pressureWeightedClosedLoopBenchmarkReport.pressureWeightedInterferenceTotal !==
      EXACT_PRESSURE_WEIGHTED_INTERFERENCE_TOTAL ||
    pressureWeightedClosedLoopBenchmarkReport.pressureUnit !==
      SIEVE_PASCAL_UNIT ||
    !sameStringTrace(
      pressureWeightedClosedLoopBenchmarkReport.theoremIds,
      PRESSURE_WEIGHTED_CLOSED_LOOP_BENCHMARK_REPORT_THEOREM_IDS
    ) ||
    interferePort.portAction !== 'interfere' ||
    interferePort.portCoordinate !== GNOSIS_URING_INTERFERE_PORT_COORDINATE ||
    interferePort.closureCoordinate !==
      GNOSIS_URING_INTERFERE_CLOSURE_COORDINATE ||
    !interferePort.terminal ||
    resolvedClosureFamilyAudit.portCoordinate !==
      GNOSIS_URING_INTERFERE_PORT_COORDINATE ||
    resolvedClosureFamilyAudit.closureFamilyPressureTotal !==
      GNOSIS_URING_BOUNDED_CLOSURE_FAMILY_PRESSURE_TOTAL ||
    resolvedClosureFamilyAudit.pressureUnit !== SIEVE_PASCAL_UNIT ||
    !sameNumberTrace(
      resolvedClosureFamilyAudit.closureIndices,
      BOUNDED_INTERFERENCE_CLOSURE_FAMILY_INDICES
    ) ||
    !sameNumberTrace(
      pressureWeightedInterference.strideTrace,
      EXACT_INTERFERENCE_STRIDE_TRACE
    ) ||
    !sameNumberTrace(
      pressureWeightedInterference.leakageTrace,
      EXACT_LEAKAGE_TRACE
    ) ||
    !sameNumberTrace(
      pressureWeightedInterference.weightedPressureTrace,
      EXACT_PRESSURE_WEIGHTED_INTERFERENCE_STRIDE_TRACE
    ) ||
    pressureWeightedInterference.weightedPressureTotal !==
      EXACT_PRESSURE_WEIGHTED_INTERFERENCE_TOTAL ||
    pressureWeightedInterference.pressureUnit !== SIEVE_PASCAL_UNIT ||
    pressureWeighted.weightedPressureTotal !==
      pressureWeightedInterference.weightedPressureTotal ||
    metrology.totalGnosticMass !== 1560 ||
    metrology.totalSievePressure !== 96 ||
    metrology.massUnit !== GNOSTIC_GRAM_UNIT ||
    metrology.pressureUnit !== SIEVE_PASCAL_UNIT
  ) {
    return null;
  }

  return {
    source: 'GnosisPressureCertifiedExecutionPathCertificate',
    certificateVersion: PRESSURE_CERTIFIED_EXECUTION_PATH_VERSION,
    pressureIdentity: PRESSURE_CERTIFIED_EXECUTION_PATH_IDENTITY,
    runtime: 'gnosis',
    host: 'gnosis-uring',
    topologyCoordinates: EXACT_CENTERLINE_GAMUT_TRACE,
    gateCoordinates: NATURAL_PORT_COORDINATES,
    decompositionCoordinates: EXACT_DECOMPOSITION_TRIPLE_COORDINATES,
    primeDoubleCrossingCount: 5,
    interferePort: {
      action: 'interfere',
      coordinate: GNOSIS_URING_INTERFERE_PORT_COORDINATE,
      closureCoordinate: GNOSIS_URING_INTERFERE_CLOSURE_COORDINATE,
      terminal: true,
    },
    pressureWeightedInterference: {
      strideTrace: EXACT_INTERFERENCE_STRIDE_TRACE,
      leakagePressureTrace: EXACT_LEAKAGE_TRACE,
      pressureWeightedStrideTrace:
        EXACT_PRESSURE_WEIGHTED_INTERFERENCE_STRIDE_TRACE,
      pressureWeightedTotal: EXACT_PRESSURE_WEIGHTED_INTERFERENCE_TOTAL,
      pressureUnit: SIEVE_PASCAL_UNIT,
    },
    closureFamilyAudit: resolvedClosureFamilyAudit,
    informationMassTotal: 1560,
    leakageTotal: 96,
    metrology: {
      informationMassTotal: 1560,
      leakageTotal: 96,
      massUnit: GNOSTIC_GRAM_UNIT,
      pressureUnit: SIEVE_PASCAL_UNIT,
    },
    pressureWeightedClosedLoopBenchmarkReport,
    returnsToBase: true,
    theoremIds: PRESSURE_CERTIFIED_EXECUTION_PATH_THEOREM_IDS,
  };
}

export interface CreateWankelRuntimeCertificateInput {
  readonly closedLoopRuntimeMetadata?: ClosedLoopRuntimeMetadataCertificate | null;
  readonly pressureWeightedClosedLoopBenchmarkReport?: PressureWeightedClosedLoopBenchmarkReportCertificate | null;
  readonly pressureCertifiedExecutionPath?: PressureCertifiedExecutionPathCertificate | null;
}

/**
 * Creates the Wankel Runtime Certificate.
 */
export function createWankelRuntimeCertificate({
  closedLoopRuntimeMetadata = createClosedLoopRuntimeMetadataCertificate(),
  pressureWeightedClosedLoopBenchmarkReport = createPressureWeightedClosedLoopBenchmarkReportCertificate(
    closedLoopRuntimeMetadata
  ),
  pressureCertifiedExecutionPath = createPressureCertifiedExecutionPathCertificate(
    { pressureWeightedClosedLoopBenchmarkReport }
  ),
}: CreateWankelRuntimeCertificateInput = {}): WankelRuntimeCertificate | null {
  if (
    closedLoopRuntimeMetadata === null ||
    pressureWeightedClosedLoopBenchmarkReport === null ||
    pressureCertifiedExecutionPath === null ||
    !returnsToBase(closedLoopRuntimeMetadata) ||
    !returnsToBase(pressureWeightedClosedLoopBenchmarkReport) ||
    !returnsToBase(pressureCertifiedExecutionPath)
  ) {
    return null;
  }

  if (
    pressureWeightedClosedLoopBenchmarkReport.closedLoopRuntimeMetadata !==
      closedLoopRuntimeMetadata ||
    pressureCertifiedExecutionPath.pressureWeightedClosedLoopBenchmarkReport !==
      pressureWeightedClosedLoopBenchmarkReport ||
    pressureCertifiedExecutionPath.runtime !== 'gnosis' ||
    pressureCertifiedExecutionPath.host !== 'gnosis-uring' ||
    !sameNumberTrace(
      pressureWeightedClosedLoopBenchmarkReport.topologyCoordinates,
      EXACT_CENTERLINE_GAMUT_TRACE
    ) ||
    !sameStringTrace(
      WANKEL_FIVE_FORCE_LIFECYCLE_TRACE,
      ['fork', 'race', 'fold', 'vent', 'interfere']
    ) ||
    WANKEL_AEON_CELLS !== 12 ||
    WANKEL_RUNTIME_DEADLINE_PULSE_HOPS !== 264
  ) {
    return null;
  }

  return {
    source: 'GnosisWankelRuntimeCertificate',
    runtime: 'gnosis',
    host: 'gnosis-uring',
    aeonCells: WANKEL_AEON_CELLS,
    lifecycleTrace: WANKEL_FIVE_FORCE_LIFECYCLE_TRACE,
    deadlinePulseHops: WANKEL_RUNTIME_DEADLINE_PULSE_HOPS,
    closedLoopRuntimeMetadata,
    pressureWeightedClosedLoopBenchmarkReport,
    pressureCertifiedExecutionPath,
    returnsToBase: true,
    theoremIds: WANKEL_RUNTIME_CERTIFICATE_THEOREM_IDS,
  };
}

/**
 * Creates the Gnostic Metrology Profile.
 */
export function createGnosticMetrologyProfile(
  certificate: StatisticalMechanicsAccountingCertificate | null = createStatisticalMechanicsAccounting()
): GnosticMetrologyProfile | null {
  if (certificate === null || !returnsToBase(certificate)) {
    return null;
  }

  const virtualMass =
    certificate.totalInformationMass - LOCAL_AEON_CARRYING_CAPACITY;
  if (virtualMass !== 1416) {
    return null;
  }

  return {
    source: 'GnosisGnosticMetrologyProfile',
    massUnit: GNOSTIC_GRAM_UNIT,
    pressureUnit: SIEVE_PASCAL_UNIT,
    localCarryingCapacity: LOCAL_AEON_CARRYING_CAPACITY,
    virtualMass,
    totalGnosticMass: certificate.totalInformationMass,
    totalSievePressure: certificate.totalLeakage,
    checksumMass: certificate.totalInformationMass,
    shootGamutPressureRequirement: certificate.totalLeakage,
    brownBufferMass: certificate.informationMassTrace[2],
    brownLeakageMass: certificate.leakageMassTrace[2],
    layerRows: [
      {
        band: EXACT_NOISE_BANDS[0],
        saturation: certificate.saturationTrace[0],
        visibleMass: certificate.visibleMassTrace[0],
        sievePressure: certificate.leakageTrace[0],
        gnosticMass: certificate.informationMassTrace[0],
        leakageMass: certificate.leakageMassTrace[0],
      },
      {
        band: EXACT_NOISE_BANDS[1],
        saturation: certificate.saturationTrace[1],
        visibleMass: certificate.visibleMassTrace[1],
        sievePressure: certificate.leakageTrace[1],
        gnosticMass: certificate.informationMassTrace[1],
        leakageMass: certificate.leakageMassTrace[1],
      },
      {
        band: EXACT_NOISE_BANDS[2],
        saturation: certificate.saturationTrace[2],
        visibleMass: certificate.visibleMassTrace[2],
        sievePressure: certificate.leakageTrace[2],
        gnosticMass: certificate.informationMassTrace[2],
        leakageMass: certificate.leakageMassTrace[2],
      },
    ],
    returnsToBase: true,
    theoremIds: certificate.theoremIds,
  };
}

/**
 * Creates the Pressure Gradient Certificate.
 */
export function createPressureGradientCertificate(
  profile: GnosticMetrologyProfile | null = createGnosticMetrologyProfile()
): PressureGradientCertificate | null {
  if (profile === null || !returnsToBase(profile)) {
    return null;
  }

  if (
    profile.totalSievePressure !== 96 ||
    PRESSURE_GRADIENT_MULTIPLIER * LOCAL_AEON_ROW_BUDGET !== 96
  ) {
    return null;
  }

  return {
    source: 'GnosisPressureGradientCertificate',
    bucketSize: LOCAL_AEON_ROW_BUDGET,
    bucketCount: PRESSURE_GRADIENT_MULTIPLIER,
    totalSievePressure: profile.totalSievePressure,
    bucketProduct: 96,
    reserveSievePressure: profile.totalSievePressure,
    returnsToBase: true,
    theoremIds: profile.theoremIds,
  };
}

/**
 * Creates the Gnostic Runtime Allocation Plan.
 */
export function createGnosticRuntimeAllocationPlan(
  observedMass: number,
  profile: GnosticMetrologyProfile | null = createGnosticMetrologyProfile()
): GnosticRuntimeAllocationPlan | null {
  if (profile === null) {
    return null;
  }
  const pressureGradient = createPressureGradientCertificate(profile);
  if (pressureGradient === null) {
    return null;
  }

  const checksumMatches =
    Number.isFinite(observedMass) && observedMass === profile.checksumMass;

  return {
    source: 'GnosisRuntimeAllocationPlan',
    observedMass,
    expectedMass: profile.checksumMass,
    checksumMatches,
    reserveBrownBufferMass: profile.brownBufferMass,
    reserveSievePressure: pressureGradient.reserveSievePressure,
    pressureGradient,
    admitted: checksumMatches,
    reason: checksumMatches ? 'checksum-match' : 'checksum-mismatch',
    theoremIds: profile.theoremIds,
  };
}

/**
 * Creates the Calibrated Fluidic Pressure Bridge.
 */
export function createCalibratedFluidicPressureBridge(
  calibrationPaPerSPa: number,
  profile: GnosticMetrologyProfile | null = createGnosticMetrologyProfile()
): CalibratedFluidicPressureBridge | null {
  if (!Number.isFinite(calibrationPaPerSPa) || calibrationPaPerSPa <= 0) {
    return null;
  }
  if (profile === null) {
    return null;
  }
  const pressureGradient = createPressureGradientCertificate(profile);
  if (pressureGradient === null) {
    return null;
  }

  const pressureTracePa: readonly [number, number, number] = [
    profile.layerRows[0].sievePressure * calibrationPaPerSPa,
    profile.layerRows[1].sievePressure * calibrationPaPerSPa,
    profile.layerRows[2].sievePressure * calibrationPaPerSPa,
  ];

  return {
    source: 'GnosisCalibratedFluidicPressureBridge',
    calibrationPaPerSPa,
    pressureTracePa,
    totalPhysicalPressurePa: profile.totalSievePressure * calibrationPaPerSPa,
    pinkPhysicalPressurePa: pressureTracePa[1],
    brownPhysicalPressurePa: pressureTracePa[2],
    pressureGradient,
    admitted: true,
    reason: 'calibrated-pressure-bridge',
    theoremIds: profile.theoremIds,
  };
}

/**
 * Creates the Fluidic Routing Admissibility Certificate.
 */
export function createFluidicRoutingAdmissibilityCertificate(
  calibrationPaPerSPa: number,
  observedMass: number,
  profile: GnosticMetrologyProfile | null = createGnosticMetrologyProfile()
): FluidicRoutingAdmissibilityCertificate | null {
  const allocationPlan = createGnosticRuntimeAllocationPlan(
    observedMass,
    profile
  );
  if (allocationPlan === null || !allocationPlan.admitted) {
    return null;
  }

  const fluidicBridge = createCalibratedFluidicPressureBridge(
    calibrationPaPerSPa,
    profile
  );
  if (fluidicBridge === null) {
    return null;
  }

  const preservesGradient =
    fluidicBridge.totalPhysicalPressurePa ===
    calibrationPaPerSPa * allocationPlan.pressureGradient.bucketProduct;
  const pressureRankPreserved =
    fluidicBridge.brownPhysicalPressurePa >=
    fluidicBridge.pinkPhysicalPressurePa;
  if (!preservesGradient || !pressureRankPreserved) {
    return null;
  }

  return {
    source: 'GnosisFluidicRoutingAdmissibilityCertificate',
    calibrationPaPerSPa,
    observedMass,
    expectedMass: allocationPlan.expectedMass,
    checksumMatches: true,
    pressureRankedBands: EXACT_PRESSURE_RANKED_BANDS,
    dominantBand: 'brown',
    dominantSievePressure: 78,
    dominantPhysicalPressurePa: fluidicBridge.brownPhysicalPressurePa,
    pinkPhysicalPressurePa: fluidicBridge.pinkPhysicalPressurePa,
    totalPhysicalPressurePa: fluidicBridge.totalPhysicalPressurePa,
    preservesGradient: true,
    allocationPlan,
    fluidicBridge,
    pressureGradient: allocationPlan.pressureGradient,
    admitted: true,
    returnsToBase: true,
    reason: 'positive-calibration-checksum-match',
    theoremIds: allocationPlan.theoremIds,
  };
}

/**
 * Handles the gnosis read Braid Fold Candidate workflow.
 */
export function readBraidFoldCandidate(
  value: unknown
): BraidFoldCandidate | null {
  if (typeof value !== 'object' || value === null) {
    return null;
  }
  const record = value as Record<string, unknown>;
  const cycle = record.cycle;
  const gateCoordinates = record.gateCoordinates;
  const saturationTrace = record.saturationTrace;
  const closedArchitecture = record.closedArchitecture;
  const endpointsLinked = record.endpointsLinked;
  const baseReady = record.baseReady;

  if (
    typeof cycle !== 'number' ||
    !isNonNegativeInteger(cycle) ||
    !Array.isArray(gateCoordinates) ||
    !Array.isArray(saturationTrace) ||
    typeof closedArchitecture !== 'boolean' ||
    typeof endpointsLinked !== 'boolean' ||
    typeof baseReady !== 'boolean'
  ) {
    return null;
  }
  if (
    !gateCoordinates.every(
      (coordinate): coordinate is number =>
        typeof coordinate === 'number' && isNonNegativeInteger(coordinate)
    ) ||
    !saturationTrace.every(
      (saturation): saturation is number =>
        typeof saturation === 'number' && isNonNegativeInteger(saturation)
    )
  ) {
    return null;
  }

  return {
    cycle,
    gateCoordinates,
    saturationTrace,
    closedArchitecture,
    endpointsLinked,
    baseReady,
  };
}

/**
 * Handles the gnosis read Decomposition Triple Candidate workflow.
 */
export function readDecompositionTripleCandidate(
  value: unknown
): DecompositionTripleCandidate | null {
  if (typeof value !== 'object' || value === null) {
    return null;
  }
  const record = value as Record<string, unknown>;
  const cycle = record.cycle;
  const rawCoordinates = record.coordinates ?? record.decompositionCoordinates;

  if (
    typeof cycle !== 'number' ||
    !isNonNegativeInteger(cycle) ||
    !Array.isArray(rawCoordinates)
  ) {
    return null;
  }
  if (
    !rawCoordinates.every(
      (coordinate): coordinate is number =>
        typeof coordinate === 'number' && isNonNegativeInteger(coordinate)
    )
  ) {
    return null;
  }

  return {
    cycle,
    coordinates: rawCoordinates,
  };
}

/**
 * Handles the gnosis audit Braid Fold Candidate workflow.
 */
export function auditBraidFoldCandidate(
  candidate: BraidFoldCandidate
): BraidFoldAudit {
  const anchorCoordinates = anchorCoordinatesFromTrace(
    candidate.gateCoordinates
  );

  if (!sameNumberTrace(candidate.gateCoordinates, EXACT_LAYER_COORDINATES)) {
    return {
      accepted: false,
      verdict: Verdict.Decline,
      reason: 'gate coordinates must be exactly 10,12,17,22',
      anchorCoordinates,
      certificate: null,
    };
  }
  if (!sameNumberTrace(candidate.saturationTrace, EXACT_SATURATION_TRACE)) {
    return {
      accepted: false,
      verdict: Verdict.Decline,
      reason: 'saturation trace must be exactly 10,30,90',
      anchorCoordinates,
      certificate: null,
    };
  }
  // The candidate is the right canonical shape (gates + saturation match) but
  // its readiness flags are not yet asserted: deliberate deferral, not a hard
  // rejection — it can be re-audited once the flags are supplied.
  if (!candidate.closedArchitecture) {
    return {
      accepted: false,
      verdict: Verdict.Abstain,
      reason: 'closed architecture certificate is absent',
      anchorCoordinates,
      certificate: null,
    };
  }
  if (!candidate.endpointsLinked) {
    return {
      accepted: false,
      verdict: Verdict.Abstain,
      reason: 'braid endpoints are not linked',
      anchorCoordinates,
      certificate: null,
    };
  }
  if (!candidate.baseReady) {
    return {
      accepted: false,
      verdict: Verdict.Abstain,
      reason: 'braid pass does not return to base readiness',
      anchorCoordinates,
      certificate: null,
    };
  }
  if (keystoneGap(17, 22) !== 5) {
    return {
      accepted: false,
      verdict: Verdict.Decline,
      reason: 'prime-to-double crossing count is not five',
      anchorCoordinates,
      certificate: null,
    };
  }

  return {
    accepted: true,
    verdict: Verdict.Accept,
    reason: 'exact braid fold certificate accepted',
    anchorCoordinates,
    certificate: {
      source: 'GnosisBraidFoldCertificate',
      cycle: candidate.cycle,
      layerCoordinates: EXACT_LAYER_COORDINATES,
      saturationTrace: EXACT_SATURATION_TRACE,
      forkedAnchorCoordinates: BRAID_FOLD_ANCHOR_COORDINATES,
      primeDoubleCrossingCount: 5,
      closedArchitecture: true,
      endpointsLinked: true,
      baseReady: true,
      returnsToBase: true,
      theoremIds: BRAID_FOLD_THEOREM_IDS,
    },
  };
}

/**
 * Creates the Braid Fold Certificate.
 */
export function createBraidFoldCertificate(
  candidate: BraidFoldCandidate
): BraidFoldCertificate | null {
  return auditBraidFoldCandidate(candidate).certificate;
}

/**
 * Handles the gnosis audit Decomposition Triple Candidate workflow.
 */
export function auditDecompositionTripleCandidate(
  candidate: DecompositionTripleCandidate,
  braidFoldCertificate: BraidFoldCertificate | null
): DecompositionTripleAudit {
  if (
    !sameNumberTrace(
      candidate.coordinates,
      EXACT_DECOMPOSITION_TRIPLE_COORDINATES
    )
  ) {
    return {
      accepted: false,
      reason: 'decomposition coordinates must be exactly 7,17,22',
      coordinates: candidate.coordinates,
      certificate: null,
    };
  }
  if (braidFoldCertificate === null) {
    return {
      accepted: false,
      reason: 'braid fold certificate is absent',
      coordinates: candidate.coordinates,
      certificate: null,
    };
  }
  if (candidate.cycle !== braidFoldCertificate.cycle) {
    return {
      accepted: false,
      reason: 'decomposition cycle must match braid fold cycle',
      coordinates: candidate.coordinates,
      certificate: null,
    };
  }

  const triple = createExactDecompositionTriple();
  if (!decompositionFeedsKenomaSpan(triple)) {
    return {
      accepted: false,
      reason: 'decomposition gate does not feed prime anchor by kenoma span',
      coordinates: candidate.coordinates,
      certificate: null,
    };
  }
  if (!decompositionAnchorsKeystones(triple, braidFoldCertificate)) {
    return {
      accepted: false,
      reason: 'decomposition anchors do not match prime and double keystones',
      coordinates: candidate.coordinates,
      certificate: null,
    };
  }
  if (
    triple.foldAnchor !==
    triple.raceAnchor + braidFoldCertificate.primeDoubleCrossingCount
  ) {
    return {
      accepted: false,
      reason: 'decomposition fold does not match prime-double crossing count',
      coordinates: candidate.coordinates,
      certificate: null,
    };
  }
  if (!returnsToBase(braidFoldCertificate)) {
    return {
      accepted: false,
      reason: 'braid fold certificate does not return to base',
      coordinates: candidate.coordinates,
      certificate: null,
    };
  }

  return {
    accepted: true,
    reason: 'exact 7,17,22 decomposition certificate accepted',
    coordinates: EXACT_DECOMPOSITION_TRIPLE_COORDINATES,
    certificate: {
      source: 'GnosisDecompositionTripleCertificate',
      cycle: candidate.cycle,
      coordinates: EXACT_DECOMPOSITION_TRIPLE_COORDINATES,
      triple: {
        decompositionGate: 7,
        raceAnchor: 17,
        foldAnchor: 22,
      },
      decompositionFeedsKenomaSpan: true,
      anchorsKeystones: true,
      primeDoubleCrossingCount: 5,
      braidFoldCertificate,
      returnsToBase: true,
      theoremIds: DECOMPOSITION_TRIPLE_THEOREM_IDS,
    },
  };
}

/**
 * Creates the Decomposition Triple Certificate.
 */
export function createDecompositionTripleCertificate(
  candidate: DecompositionTripleCandidate,
  braidFoldCertificate: BraidFoldCertificate | null
): DecompositionTripleCertificate | null {
  return auditDecompositionTripleCandidate(candidate, braidFoldCertificate)
    .certificate;
}

/**
 * Handles the gnosis select Certified Braid Fast Path workflow.
 */
export function selectCertifiedBraidFastPath(
  certificate: BraidFoldCertificate | null
): CertifiedBraidFastPathDecision {
  if (certificate === null) {
    return {
      selected: false,
      verdict: Verdict.Decline,
      reason: 'braid fold certificate is absent',
      path: null,
    };
  }
  if (!returnsToBase(certificate)) {
    // A present certificate that we deliberately do not adopt because the
    // base-return gate is unmet: deferred non-adoption, not a hard rejection.
    return {
      selected: false,
      verdict: Verdict.Abstain,
      reason: 'braid fold certificate does not return to base',
      path: null,
    };
  }

  return {
    selected: true,
    verdict: Verdict.Accept,
    reason: 'certified braid fast path selected',
    path: {
      source: 'GnosisCertifiedBraidFastPath',
      certificate,
      forkedAnchorCoordinates: BRAID_FOLD_ANCHOR_COORDINATES,
      collapsedBridge: {
        from: 17,
        to: 22,
        stride: 5,
      },
      returnsToBase: true,
    },
  };
}

export function certifyBraidFastPathEquivalence<T>(
  certificate: BraidFoldCertificate | null,
  linearResult: T,
  certifiedResult: T,
  equivalent: (linearResult: T, certifiedResult: T) => boolean = Object.is
): CertifiedBraidFastPathEquivalence<T> {
  const decision = selectCertifiedBraidFastPath(certificate);
  return {
    decision,
    equivalent: decision.selected && equivalent(linearResult, certifiedResult),
    linearResult,
    certifiedResult,
  };
}

function isExactWankelRuntimeCertificate(
  certificate: WankelRuntimeCertificate
): boolean {
  return (
    certificate.source === 'GnosisWankelRuntimeCertificate' &&
    certificate.runtime === 'gnosis' &&
    certificate.host === 'gnosis-uring' &&
    certificate.aeonCells === WANKEL_AEON_CELLS &&
    certificate.deadlinePulseHops === WANKEL_RUNTIME_DEADLINE_PULSE_HOPS &&
    sameStringTrace(
      certificate.lifecycleTrace,
      WANKEL_FIVE_FORCE_LIFECYCLE_TRACE
    ) &&
    certificate.pressureWeightedClosedLoopBenchmarkReport
      .closedLoopRuntimeMetadata === certificate.closedLoopRuntimeMetadata &&
    certificate.pressureCertifiedExecutionPath
      .pressureWeightedClosedLoopBenchmarkReport ===
      certificate.pressureWeightedClosedLoopBenchmarkReport &&
    returnsToBase(certificate.closedLoopRuntimeMetadata) &&
    returnsToBase(certificate.pressureWeightedClosedLoopBenchmarkReport) &&
    returnsToBase(certificate.pressureCertifiedExecutionPath) &&
    sameStringTrace(
      certificate.theoremIds,
      WANKEL_RUNTIME_CERTIFICATE_THEOREM_IDS
    )
  );
}

/**
 * Handles the gnosis select Wankel Runtime Fast Path workflow.
 */
export function selectWankelRuntimeFastPath(
  certificate: WankelRuntimeCertificate | null
): WankelRuntimeFastPathDecision {
  if (certificate === null) {
    return {
      selected: false,
      verdict: Verdict.Decline,
      reason: 'wankel runtime certificate is absent',
      certificate: null,
      deadlinePulseHops: null,
      lifecycleTrace: null,
    };
  }
  if (!returnsToBase(certificate)) {
    // A present certificate deliberately not adopted because the base-return
    // gate is unmet: deferred non-adoption, not a malformed candidate.
    return {
      selected: false,
      verdict: Verdict.Abstain,
      reason: 'wankel runtime certificate does not return to base',
      certificate: null,
      deadlinePulseHops: null,
      lifecycleTrace: null,
    };
  }
  if (!isExactWankelRuntimeCertificate(certificate)) {
    // Wrong shape: an inexact certificate can never be the Wankel fast path —
    // a hard rejection of a malformed candidate.
    return {
      selected: false,
      verdict: Verdict.Decline,
      reason: 'wankel runtime certificate is inexact',
      certificate: null,
      deadlinePulseHops: null,
      lifecycleTrace: null,
    };
  }

  return {
    selected: true,
    verdict: Verdict.Accept,
    reason: 'wankel runtime fast path selected',
    certificate,
    deadlinePulseHops: WANKEL_RUNTIME_DEADLINE_PULSE_HOPS,
    lifecycleTrace: WANKEL_FIVE_FORCE_LIFECYCLE_TRACE,
  };
}

export function certifyWankelRuntimeFastPathEquivalence<T>(
  certificate: WankelRuntimeCertificate | null,
  linearResult: T,
  certifiedResult: T,
  equivalent: (linearResult: T, certifiedResult: T) => boolean = Object.is
): WankelRuntimeFastPathEquivalence<T> {
  const decision = selectWankelRuntimeFastPath(certificate);
  return {
    decision,
    equivalent: decision.selected && equivalent(linearResult, certifiedResult),
    linearResult,
    certifiedResult,
  };
}

export function certifyWankelRuntimeFastPathCostDominance<T>(
  certificate: WankelRuntimeCertificate | null,
  linearResult: T,
  certifiedResult: T,
  witness: WankelRuntimeFastPathCostWitness,
  equivalent: (linearResult: T, certifiedResult: T) => boolean = Object.is
): WankelRuntimeFastPathCostDominance<T> {
  const equivalence = certifyWankelRuntimeFastPathEquivalence(
    certificate,
    linearResult,
    certifiedResult,
    equivalent
  );
  const finitePositiveCosts =
    Number.isFinite(witness.selectedCost) &&
    Number.isFinite(witness.baselineCost) &&
    witness.selectedCost > 0 &&
    witness.baselineCost > 0;
  const dominates =
    equivalence.equivalent &&
    finitePositiveCosts &&
    witness.selectedCost < witness.baselineCost;

  return {
    equivalence,
    witness,
    dominates,
    reason: !equivalence.decision.selected
      ? equivalence.decision.reason
      : !equivalence.equivalent
        ? 'wankel runtime fast path result is not equivalent'
        : !finitePositiveCosts
          ? 'wankel runtime cost witness must be positive and finite'
          : dominates
            ? 'wankel runtime fast path cost dominance certified'
            : 'wankel runtime fast path is not faster than baseline',
    theoremId: 'RuntimeCertificate.wankel_fast_path_cost_dominates_baseline',
  };
}

export function certifyWankelRuntimeFastPathAggregateCostDominance<T>(
  certificate: WankelRuntimeCertificate | null,
  linearResult: T,
  certifiedResult: T,
  witness: WankelRuntimeFastPathAggregateCostWitness,
  equivalent: (linearResult: T, certifiedResult: T) => boolean = Object.is
): WankelRuntimeFastPathAggregateCostDominance<T> {
  const equivalence = certifyWankelRuntimeFastPathEquivalence(
    certificate,
    linearResult,
    certifiedResult,
    equivalent
  );
  const validSampleCount =
    Number.isInteger(witness.sampleCount) && witness.sampleCount > 0;
  const finitePositiveCosts =
    Number.isFinite(witness.selectedTotalCost) &&
    Number.isFinite(witness.baselineTotalCost) &&
    witness.selectedTotalCost > 0 &&
    witness.baselineTotalCost > 0;
  const dominates =
    equivalence.equivalent &&
    validSampleCount &&
    finitePositiveCosts &&
    witness.selectedTotalCost < witness.baselineTotalCost;

  return {
    equivalence,
    witness,
    dominates,
    reason: !equivalence.decision.selected
      ? equivalence.decision.reason
      : !equivalence.equivalent
        ? 'wankel runtime fast path result is not equivalent'
        : !validSampleCount
          ? 'wankel runtime aggregate cost witness sample count must be a positive integer'
          : !finitePositiveCosts
            ? 'wankel runtime aggregate costs must be positive and finite'
            : dominates
              ? 'wankel runtime aggregate fast path cost dominance certified'
              : 'wankel runtime aggregate fast path is not faster than baseline',
    theoremId:
      'RuntimeCertificate.wankel_aggregate_fast_path_cost_dominates_baseline',
  };
}

export function certifyWankelRuntimeFastPathRobustEstimatorDominance<T>(
  certificate: WankelRuntimeCertificate | null,
  linearResult: T,
  certifiedResult: T,
  witness: WankelRuntimeFastPathRobustEstimatorWitness,
  equivalent: (linearResult: T, certifiedResult: T) => boolean = Object.is
): WankelRuntimeFastPathRobustEstimatorDominance<T> {
  const equivalence = certifyWankelRuntimeFastPathEquivalence(
    certificate,
    linearResult,
    certifiedResult,
    equivalent
  );
  const validSampleCoverage =
    Number.isInteger(witness.sampleCount) &&
    Number.isInteger(witness.minimumSampleCount) &&
    witness.minimumSampleCount > 0 &&
    witness.sampleCount >= witness.minimumSampleCount;
  const finitePositiveAggregateCosts =
    Number.isFinite(witness.selectedTotalCost) &&
    Number.isFinite(witness.baselineTotalCost) &&
    witness.selectedTotalCost > 0 &&
    witness.baselineTotalCost > 0;
  const finitePositiveRobustCosts =
    Number.isFinite(witness.selectedRobustCost) &&
    Number.isFinite(witness.baselineRobustCost) &&
    witness.selectedRobustCost > 0 &&
    witness.baselineRobustCost > 0;
  const validJitterBudget =
    Number.isFinite(witness.observedJitter) &&
    Number.isFinite(witness.jitterBudget) &&
    witness.observedJitter >= 0 &&
    witness.jitterBudget >= 0 &&
    witness.observedJitter <= witness.jitterBudget;
  const validOutlierBudget =
    Number.isInteger(witness.outlierBudget) &&
    witness.outlierBudget >= 0 &&
    Number.isInteger(witness.sampleCount) &&
    witness.outlierBudget < witness.sampleCount;
  const dominates =
    equivalence.equivalent &&
    validSampleCoverage &&
    finitePositiveAggregateCosts &&
    witness.selectedTotalCost < witness.baselineTotalCost &&
    finitePositiveRobustCosts &&
    witness.selectedRobustCost < witness.baselineRobustCost &&
    validJitterBudget &&
    validOutlierBudget;

  return {
    equivalence,
    witness,
    dominates,
    reason: !equivalence.decision.selected
      ? equivalence.decision.reason
      : !equivalence.equivalent
        ? 'wankel runtime fast path result is not equivalent'
        : !validSampleCoverage
          ? 'wankel runtime robust estimator witness requires minimum sample coverage'
          : !finitePositiveAggregateCosts
            ? 'wankel runtime robust estimator aggregate costs must be positive and finite'
            : witness.selectedTotalCost >= witness.baselineTotalCost
              ? 'wankel runtime robust estimator aggregate fast path is not faster than baseline'
              : !finitePositiveRobustCosts
                ? 'wankel runtime robust estimator costs must be positive and finite'
                : witness.selectedRobustCost >= witness.baselineRobustCost
                  ? 'wankel runtime robust estimator fast path is not faster than baseline'
                  : !validJitterBudget
                    ? 'wankel runtime robust estimator jitter exceeds budget'
                    : !validOutlierBudget
                      ? 'wankel runtime robust estimator outlier budget must be within samples'
                      : 'wankel runtime robust estimator cost dominance certified',
    theoremId:
      'RuntimeCertificate.wankel_robust_fast_path_estimator_dominates_baseline',
  };
}
