// @a0n/gnosis/src/mesh-policy-edge.ts
/**
 * Mesh-policy edge runtime — Web Fetch pure.
 *
 * A Node-free extraction of x-ranger's runtime mesh-policy middleware
 * into the gnosis package. Runs on any host that exposes the Web Fetch
 * API surface (`Request`, `Response`, `Headers`, `URL`): Cloudflare
 * Workers, Deno, Bun, Node 18+, browsers, edge functions.
 *
 * Zero dependencies on Node-specific modules: no `node:*` imports, no
 * `process`, no `console`, no `Buffer`, no `fs`. Telemetry is plumbed
 * via an optional `audit` callback on {@link WrapOptions}; the runtime
 * is silent when no callback is supplied.
 *
 * Formal primitives are sourced from `@a0n/gnosis/mesh-policy`
 * (the zero-sorry Lean contract mirror):
 *   - `Mesh`, `PortOverlay`, `createMesh`, `createPortOverlay`
 *   - `meshNetHitCountWithOverlay`
 *   - `Link`, `createLink`
 *   - ICE/anti-ICE bypass witnesses: `overlayHoleBypass`,
 *     `clockSideChannelBypass`, `zenonicApproachBypass`
 *
 * Relevant Lean theorems that shape the edge gate contract:
 *   - `emptyPortListSealsRegion`         — default-deny ingress (404)
 *   - `portOverlayCancelsWithinPorts`    — overlay cancellation (403)
 *   - `authenticationRestrictsPorts`     — auth-header presence gate (401)
 *   - `fullPortListEqualsOpenMesh`       — delegate (200)
 *   - `linkRequiresSourcePort`           — linked-mesh admission
 *   - `linkRequiresDestPort`             — linked-mesh admission
 *   - `linkHonorsSourceOverlay`          — linked-mesh overlay gate
 *
 * This file intentionally merges portions of three x-ranger files
 * (`mesh-policy.ts`, `mesh-audit.ts`, `linked-mesh.ts`) into a single
 * edge module. Honey-pit pool state is NOT included here; consumers that
 * need trap routing plumb it via {@link WrapOptions.overlayHandler}.
 */

import {
  clockSideChannelBypass,
  createClock,
  createGap,
  createLink,
  createMesh,
  createOverlay,
  createPortOverlay,
  meshNetHitCountWithOverlay,
  overlayHoleBypass,
  zenonicApproachBypass,
  type Link,
  type Mesh,
  type PortOverlay,
} from './mesh-policy.js';

// ─── Public config types (x-ranger-agnostic) ─────────────────────────

export interface MeshPortSpec {
  readonly path: string;
  readonly methods?: readonly string[];
}

export interface MeshOverlaySpec {
  readonly pathPrefix: string;
  readonly reason?: string;
  /**
   * Flag only — this edge package does NOT handle honey-pit trapping.
   * Consumers that want custom overlay responses should use
   * {@link WrapOptions.overlayHandler}.
   */
  readonly honeypit?: boolean;
}

export interface MeshHoneyPitConfig {
  readonly capacity: number;
  readonly drainSlots: number;
  /**
   * Backing store for the pool's `inhabitants` counter + drain ticks.
   *
   * - `'memory'` (default) — module-level counter inside the wrapping
   *   Worker isolate. Cheap, but state resets every time Cloudflare
   *   recycles the isolate, so the cross-Worker aikido bound is only
   *   observational (an attacker that forces a fresh isolate gets a
   *   fresh pool).
   * - `'do'` — keyed per-policy Durable Object. Storage survives
   *   isolate recycling so `HoneyPit.sharedDrainInvariant` and
   *   `RateLimitBarrier.rateLimitPreservesAikidoServerBound` hold
   *   across the whole deployment, not just within one isolate.
   *
   * When `'do'`, {@link doBindingName} must point to a
   * `DurableObjectNamespace` binding on `env`.
   */
  readonly storage?: 'memory' | 'do';
  /**
   * Name of the `env[...]` binding that resolves to the
   * `DurableObjectNamespace` hosting the
   * `MeshHoneyPitDurableObject` class.
   *
   * Required when {@link storage} === `'do'`; ignored otherwise.
   */
  readonly doBindingName?: string;
}

export interface LinkedServiceMesh {
  readonly sourcePortIdx: number;
  readonly destPortIdx: number;
  readonly inheritOverlays: boolean;
}

export interface MeshServiceBinding {
  readonly binding: string;
  readonly service: string;
  readonly mesh?: LinkedServiceMesh;
}

export type AdoptionMode = 'audit' | 'enforce' | 'shadow';

export interface MeshPolicyConfig {
  readonly version?: number;
  readonly ports: readonly MeshPortSpec[];
  /**
   * DEPRECATED alias for {@link MeshPolicyConfig.requireAuthHeaderPresent}.
   * The runtime reads `requireAuthHeaderPresent ?? requireAuth`.
   */
  readonly requireAuth?: boolean;
  /**
   * Canonical auth-gate flag: presence-only. When true, the edge wrapper
   * returns 401 if no `Authorization` header is present.
   */
  readonly requireAuthHeaderPresent?: boolean;
  readonly overlays?: readonly MeshOverlaySpec[];
  readonly honeypit?: MeshHoneyPitConfig;
  readonly mode?: AdoptionMode;
}

// ─── Path matchers ───────────────────────────────────────────────────

/**
 * True iff `path` exactly equals `prefix` or begins with `prefix + '/'`.
 * Segment-boundary predicate that prevents the classic `startsWith`
 * prefix-boundary bug where `/api` accidentally matches `/apiary`.
 *
 * Trailing-slash on prefix is normalized: `/api/` and `/api` are
 * equivalent prefixes.
 */
export function pathMatchesPrefix(path: string, prefix: string): boolean {
  const normalized = prefix.endsWith('/') ? prefix.slice(0, -1) : prefix;
  if (path === normalized) return true;
  return path.startsWith(normalized + '/');
}

/** Backward-compatible alias: port-path matching uses the same
 *  segment-boundary predicate as overlay matching. */
export function pathMatchesPort(path: string, portPath: string): boolean {
  return pathMatchesPrefix(path, portPath);
}

/** Sanitize an overlay reason for echoing into a 403 body.
 *
 *  - Replaces every ASCII control character (C0 range 0x00-0x1F and DEL
 *    0x7F) with a single space.
 *  - Caps the result at 200 characters; longer strings are truncated
 *    and suffixed with `"..."`.
 *  - Empty post-trim falls back to `"(no reason)"`.
 */
export function sanitizeReason(raw: string): string {
  // eslint-disable-next-line no-control-regex
  const stripped = raw.replace(/[\u0000-\u001F\u007F]/g, ' ').trim();
  if (stripped.length === 0) return '(no reason)';
  if (stripped.length > 200) return stripped.slice(0, 200) + '...';
  return stripped;
}

// ─── Wrap options and inner fetch ────────────────────────────────────

export type InnerFetch = (
  request: Request,
  env: Record<string, unknown>,
  ctx: unknown,
) => Response | Promise<Response>;

export type WrappedFetch = (
  request: Request,
  env: Record<string, unknown>,
  ctx: unknown,
) => Promise<Response>;

/**
 * Gate identifier for observability events.
 */
export type GateName = 'overlay' | 'path' | 'method' | 'auth' | 'delegate';

/**
 * Structured event delivered to {@link WrapOptions.onGateDecision}
 * once per gate evaluation.
 *
 * The `status` field is a placeholder for the `'delegate'` event
 * (always 200); the actual inner-fetch status is not observed here.
 */
export interface GateDecisionEvent {
  readonly gate: GateName;
  readonly decision: 'allow' | 'deny';
  readonly status: number;
  readonly path: string;
  readonly method: string;
  readonly reason?: string;
  readonly theorem: string;
  readonly matchedPortPath?: string;
  readonly matchedOverlayPrefix?: string;
  readonly methodsAllowed?: readonly string[];
  readonly authHeaderPresent?: boolean;
  readonly timestampMs: number;
}

export interface WrapOptions {
  /**
   * Telemetry callback. Invoked by audit/shadow modes for every gate
   * decision; never invoked in enforce mode (the standard path).
   * When absent, the runtime is silent.
   */
  readonly audit?: (message: string) => void;
  /**
   * Overlay response hook. When an overlay matches, the consumer may
   * return a custom `Response` (for example a honey-pit slow-trickle
   * body). Returning `null` — synchronously or asynchronously — falls
   * through to the standard 403 reply. Absent hook is equivalent to
   * always returning `null`.
   */
  readonly overlayHandler?: (
    overlay: MeshOverlaySpec,
    request: Request,
  ) => Response | Promise<Response | null> | null;
  /**
   * Optional async auth verifier. Invoked AFTER the presence check on
   * gate 4, only when `require_auth` (or `requireAuthHeaderPresent`) is
   * true AND the request has a non-empty Authorization header. If the
   * verifier returns false OR throws, the request is denied with 401.
   *
   * Callers: JWT decoder, UCAN validator, API-key lookup, or any custom
   * predicate. Return a resolved Promise; the runtime awaits it.
   *
   * Cites: `authenticationRestrictsPorts` (formal gate 4 contract).
   * This hook refines the presence-only check into structured validation
   * without altering the gate ordering.
   */
  readonly authVerifier?: (
    request: Request,
    authHeader: string,
  ) => Promise<boolean>;
  /**
   * Optional callback fired at every gate decision (deny or allow).
   * Used by consumers to log to audit stores, emit metrics, build
   * observability dashboards, etc. Runs synchronously; errors from
   * the callback are caught and discarded (never surface in response).
   */
  readonly onGateDecision?: (event: GateDecisionEvent) => void;
}

/**
 * Invoke an observability callback without letting its errors escape.
 * A broken observer must never affect the gate decision or response.
 */
function safelyEmitGateDecision(
  fn: WrapOptions['onGateDecision'],
  event: GateDecisionEvent,
): void {
  if (!fn) return;
  try {
    fn(event);
  } catch {
    // swallow; observer errors must not affect response
  }
}

// ─── Compiled-policy internals ───────────────────────────────────────

interface CompiledOverlay {
  readonly spec: MeshOverlaySpec;
}

interface CompiledPolicy {
  readonly overlays: readonly CompiledOverlay[];
  readonly portPaths: readonly string[];
  readonly portMethods: Map<string, ReadonlySet<string> | null>;
  readonly portByPath: Map<string, MeshPortSpec>;
  readonly requireAuthHeader: boolean;
  readonly mode: AdoptionMode;
}

function compilePolicy(policy: MeshPolicyConfig): CompiledPolicy {
  // Longest-prefix-first so the first overlay match is the most
  // specific one.
  const overlays: CompiledOverlay[] = (policy.overlays ?? [])
    .slice()
    .sort((a, b) => b.pathPrefix.length - a.pathPrefix.length)
    .map((spec) => ({ spec }));

  const portPaths: string[] = policy.ports
    .map((p) => p.path)
    .slice()
    .sort((a, b) => b.length - a.length);

  const portByPath = new Map<string, MeshPortSpec>();
  const portMethods = new Map<string, ReadonlySet<string> | null>();
  for (const port of policy.ports) {
    portByPath.set(port.path, port);
    if (port.methods && port.methods.length > 0) {
      const set = new Set<string>();
      for (const m of port.methods) set.add(m.toUpperCase());
      portMethods.set(port.path, set);
    } else {
      portMethods.set(port.path, null);
    }
  }

  const requireAuthHeader =
    (policy.requireAuthHeaderPresent ?? policy.requireAuth) === true;

  const rawMode = policy.mode;
  const mode: AdoptionMode =
    rawMode === 'audit' || rawMode === 'shadow' ? rawMode : 'enforce';

  return {
    overlays,
    portPaths,
    portMethods,
    portByPath,
    requireAuthHeader,
    mode,
  };
}

function findCompiledBlockingOverlay(
  path: string,
  overlays: readonly CompiledOverlay[],
): MeshOverlaySpec | undefined {
  for (const overlay of overlays) {
    if (pathMatchesPrefix(path, overlay.spec.pathPrefix)) return overlay.spec;
  }
  return undefined;
}

function findCompiledPortPath(
  path: string,
  portPaths: readonly string[],
): string | undefined {
  for (const portPath of portPaths) {
    if (pathMatchesPort(path, portPath)) return portPath;
  }
  return undefined;
}

/**
 * Common headers applied to every 4xx mesh-policy response.
 *
 * Per RFC 9110 §5.2 and §13.2, error responses that carry overlay
 * reason info (for example a 403 body echoing the operator-supplied
 * overlay `reason`) must not be cached by intermediaries: the denial
 * is specific to the requesting identity and context. `Cache-Control:
 * no-store` is the strongest directive, prohibiting any cache from
 * retaining the response, including private browser caches.
 */
const MESH_POLICY_4XX_COMMON_HEADERS: Record<string, string> = {
  'content-type': 'text/plain; charset=utf-8',
  'cache-control': 'no-store',
};

function textResponse(
  status: number,
  body: string,
  extraHeaders?: Record<string, string>,
): Response {
  const headers: Record<string, string> = {
    ...MESH_POLICY_4XX_COMMON_HEADERS,
  };
  if (extraHeaders) {
    for (const key of Object.keys(extraHeaders)) {
      headers[key] = extraHeaders[key] as string;
    }
  }
  return new Response(body, { status, headers });
}

/**
 * Content-negotiated JSON body kinds for 4xx mesh-policy responses.
 */
type JsonErrorKind =
  | 'overlay-blocked'
  | 'path-not-whitelisted'
  | 'method-not-allowed'
  | 'auth-required'
  | 'internal';

interface JsonErrorBody {
  readonly error: JsonErrorKind;
  readonly status: number;
  readonly path: string;
  readonly method: string;
  readonly theorem: string;
  readonly reason?: string;
  readonly matchedOverlay?: string;
  readonly methodsAllowed?: readonly string[];
}

function jsonResponse(
  status: number,
  body: JsonErrorBody,
  extraHeaders?: Record<string, string>,
): Response {
  const headers: Record<string, string> = {
    'content-type': 'application/json; charset=utf-8',
    'cache-control': 'no-store',
  };
  if (extraHeaders) {
    for (const key of Object.keys(extraHeaders)) {
      headers[key] = extraHeaders[key] as string;
    }
  }
  return new Response(JSON.stringify(body), { status, headers });
}

/**
 * True iff the request's `Accept` header indicates the caller prefers
 * JSON for the response body. Uses the same loose `includes` check as
 * documented in the content-negotiation contract.
 */
function acceptsJson(request: Request): boolean {
  const accept = request.headers.get('accept');
  return accept !== null && accept.includes('application/json');
}

// ─── wrapWithMeshPolicy ──────────────────────────────────────────────

/**
 * Wrap an inner Web Fetch handler with mesh-policy enforcement.
 *
 * Gate ordering (matches x-ranger exactly):
 *   1. Overlay cancellation (`portOverlayCancelsWithinPorts`).
 *   2. Default-deny ingress (`emptyPortListSealsRegion`).
 *   3. Method filter on matched port.
 *   4. Authentication-header-presence gate
 *      (`authenticationRestrictsPorts`) — presence-only, no token
 *      semantics.
 *   5. Delegate to `innerFetch`.
 *
 * Adoption mode (`policy.mode`):
 *   - `'enforce'` (default): gate decisions produce 4xx responses.
 *   - `'audit'`: gate decisions are computed and forwarded to the
 *     {@link WrapOptions.audit} callback, but the inner fetch is
 *     delegated regardless.
 *   - `'shadow'`: same as audit; different log prefix.
 *
 * The returned handler is purely Web-Fetch based and allocates only
 * the outgoing `Response` on the hot path.
 */
export function wrapWithMeshPolicy(
  innerFetch: InnerFetch,
  policy: MeshPolicyConfig,
  options?: WrapOptions,
): WrappedFetch {
  const compiled = compilePolicy(policy);
  const {
    overlays,
    portPaths,
    portMethods,
    portByPath,
    requireAuthHeader,
    mode,
  } = compiled;

  const auditLogPrefix =
    mode === 'audit'
      ? '[mesh-policy-edge:audit]'
      : mode === 'shadow'
        ? '[mesh-policy-edge:shadow]'
        : null;

  const audit = options?.audit;
  const overlayHandler = options?.overlayHandler;
  const authVerifier = options?.authVerifier;
  const onGateDecision = options?.onGateDecision;

  interface DenyContext {
    readonly req: Request;
    readonly env: Record<string, unknown>;
    readonly ctx: unknown;
    readonly gate: GateName;
    readonly status: number;
    readonly textBody: string;
    readonly jsonKind: JsonErrorKind;
    readonly extraHeaders: Record<string, string> | undefined;
    readonly path: string;
    readonly method: string;
    readonly theorem: string;
    readonly reason?: string;
    readonly matchedPortPath?: string;
    readonly matchedOverlayPrefix?: string;
    readonly methodsAllowed?: readonly string[];
    readonly authHeaderPresent?: boolean;
  }

  const maybeBlock = async (d: DenyContext): Promise<Response> => {
    const event: GateDecisionEvent = {
      gate: d.gate,
      decision: 'deny',
      status: d.status,
      path: d.path,
      method: d.method,
      theorem: d.theorem,
      timestampMs: Date.now(),
      ...(d.reason !== undefined ? { reason: d.reason } : {}),
      ...(d.matchedPortPath !== undefined
        ? { matchedPortPath: d.matchedPortPath }
        : {}),
      ...(d.matchedOverlayPrefix !== undefined
        ? { matchedOverlayPrefix: d.matchedOverlayPrefix }
        : {}),
      ...(d.methodsAllowed !== undefined
        ? { methodsAllowed: d.methodsAllowed }
        : {}),
      ...(d.authHeaderPresent !== undefined
        ? { authHeaderPresent: d.authHeaderPresent }
        : {}),
    };
    safelyEmitGateDecision(onGateDecision, event);

    if (auditLogPrefix !== null) {
      if (audit) {
        audit(
          `${auditLogPrefix} gate=${d.gate} status=${d.status} path=${d.path} method=${d.method}`,
        );
      }
      return await innerFetch(d.req, d.env, d.ctx);
    }

    if (acceptsJson(d.req)) {
      const jsonBody: JsonErrorBody = {
        error: d.jsonKind,
        status: d.status,
        path: d.path,
        method: d.method,
        theorem: d.theorem,
        ...(d.reason !== undefined ? { reason: d.reason } : {}),
        ...(d.matchedOverlayPrefix !== undefined
          ? { matchedOverlay: d.matchedOverlayPrefix }
          : {}),
        ...(d.methodsAllowed !== undefined
          ? { methodsAllowed: d.methodsAllowed }
          : {}),
      };
      return jsonResponse(d.status, jsonBody, d.extraHeaders);
    }

    return textResponse(d.status, d.textBody, d.extraHeaders);
  };

  return async (req, env, ctx) => {
    const url = new URL(req.url);
    const path = url.pathname;
    const methodRequested = req.method.toUpperCase();

    // Gate 1: overlay cancellation.
    const blockingOverlay = findCompiledBlockingOverlay(path, overlays);
    if (blockingOverlay) {
      // Consumer-supplied overlay hook (for example, honey-pit routing).
      // Only consulted in enforce mode. Audit/shadow modes delegate
      // through to the inner fetch via maybeBlock per x-ranger parity.
      if (auditLogPrefix === null && overlayHandler) {
        const handled = await overlayHandler(blockingOverlay, req);
        if (handled !== null) {
          return handled;
        }
      }
      const rawReason = blockingOverlay.reason ?? 'unspecified';
      const reason = sanitizeReason(rawReason);
      const textBody =
        `mesh-policy-edge: overlay-blocked: ${reason}; ` +
        `path=${path}; matchedOverlay=${blockingOverlay.pathPrefix}; reason=${reason}`;
      return await maybeBlock({
        req,
        env,
        ctx,
        gate: 'overlay',
        status: 403,
        textBody,
        jsonKind: 'overlay-blocked',
        extraHeaders: undefined,
        path,
        method: methodRequested,
        theorem: 'portOverlayCancelsWithinPorts',
        reason,
        matchedOverlayPrefix: blockingOverlay.pathPrefix,
      });
    }

    // Gate 2: default-deny ingress.
    const matchedPortPath = findCompiledPortPath(path, portPaths);
    if (!matchedPortPath) {
      const textBody =
        `mesh-policy-edge: path not on port whitelist; ` +
        `path=${path}; methodRequested=${methodRequested}`;
      return await maybeBlock({
        req,
        env,
        ctx,
        gate: 'path',
        status: 404,
        textBody,
        jsonKind: 'path-not-whitelisted',
        extraHeaders: undefined,
        path,
        method: methodRequested,
        theorem: 'deploymentShapesToContainer',
        reason: 'path not on port whitelist',
      });
    }

    const matchedPort = portByPath.get(matchedPortPath) as MeshPortSpec;
    const allowedMethods = portMethods.get(matchedPortPath) ?? null;

    // OPTIONS: administrative — respond 200 with Allow header.
    if (methodRequested === 'OPTIONS') {
      const methodList = allowedMethods
        ? Array.from(allowedMethods.values())
        : ['*'];
      const allowHeader = [...methodList, 'OPTIONS'].join(', ');
      return new Response(null, {
        status: 200,
        headers: {
          Allow: allowHeader,
          'content-type': 'text/plain; charset=utf-8',
        },
      });
    }

    // Gate 3: method filter. HEAD auto-allow when GET is in the set.
    if (allowedMethods !== null) {
      let methodOk = allowedMethods.has(methodRequested);
      if (
        !methodOk &&
        methodRequested === 'HEAD' &&
        allowedMethods.has('GET')
      ) {
        methodOk = true;
      }
      if (!methodOk) {
        const methodsAllowedList = Array.from(allowedMethods.values());
        const methodsAllowed = methodsAllowedList.join(',');
        const textBody =
          `mesh-policy-edge: method not allowed on port; ` +
          `path=${path}; methodRequested=${methodRequested}; methodsAllowed=${methodsAllowed}`;
        const headers: Record<string, string> = {
          Allow: methodsAllowedList.join(', '),
        };
        return await maybeBlock({
          req,
          env,
          ctx,
          gate: 'method',
          status: 405,
          textBody,
          jsonKind: 'method-not-allowed',
          extraHeaders: headers,
          path,
          method: methodRequested,
          theorem: 'portOverlayCancelsWithinPorts',
          reason: `method not allowed on port '${matchedPortPath}'`,
          matchedPortPath,
          methodsAllowed: methodsAllowedList,
        });
      }
    }

    // Gate 4: auth-header-presence (plus optional structured verifier).
    if (requireAuthHeader) {
      const authHeader = req.headers.get('authorization');
      if (!authHeader) {
        const textBody = `mesh-policy-edge: authentication required; path=${path}`;
        return await maybeBlock({
          req,
          env,
          ctx,
          gate: 'auth',
          status: 401,
          textBody,
          jsonKind: 'auth-required',
          extraHeaders: { 'WWW-Authenticate': 'Bearer realm="mesh-policy"' },
          path,
          method: methodRequested,
          theorem: 'authenticationRestrictsPorts',
          reason: 'authentication required',
          matchedPortPath,
          authHeaderPresent: false,
        });
      }
      if (authVerifier) {
        let verifierOk = false;
        try {
          verifierOk = await authVerifier(req, authHeader);
        } catch (err) {
          // Fail-closed on verifier error. NEVER leak error detail into
          // the response body; the `audit` callback (if any) receives a
          // structured log message for operators.
          if (audit) {
            const detail = err instanceof Error ? err.message : String(err);
            audit(`authVerifier threw: ${detail}`);
          }
          const textBody = `mesh-policy-edge: authentication required; path=${path}`;
          return await maybeBlock({
            req,
            env,
            ctx,
            gate: 'auth',
            status: 401,
            textBody,
            jsonKind: 'auth-required',
            extraHeaders: { 'WWW-Authenticate': 'Bearer realm="mesh-policy"' },
            path,
            method: methodRequested,
            theorem: 'authenticationRestrictsPorts',
            reason: 'authentication required',
            matchedPortPath,
            authHeaderPresent: true,
          });
        }
        if (!verifierOk) {
          const textBody = `mesh-policy-edge: authentication required; path=${path}`;
          return await maybeBlock({
            req,
            env,
            ctx,
            gate: 'auth',
            status: 401,
            textBody,
            jsonKind: 'auth-required',
            extraHeaders: { 'WWW-Authenticate': 'Bearer realm="mesh-policy"' },
            path,
            method: methodRequested,
            theorem: 'authenticationRestrictsPorts',
            reason: 'authentication required',
            matchedPortPath,
            authHeaderPresent: true,
          });
        }
      }
    }

    // Satisfy the TS reader: matchedPort is intentionally unused in the
    // hot path (per-port metadata lookup is reserved for future gates).
    void matchedPort;

    // Gate 5: delegate. Fire an allow event BEFORE invoking the inner
    // fetch. The `status: 200` here is a gate placeholder meaning
    // "gate allows this to reach the inner worker"; the actual inner
    // status is the worker's responsibility and not observed here.
    const authHeaderPresentForDelegate = requireAuthHeader
      ? req.headers.get('authorization') !== null
      : undefined;
    safelyEmitGateDecision(onGateDecision, {
      gate: 'delegate',
      decision: 'allow',
      status: 200,
      path,
      method: methodRequested,
      theorem: 'fullPortListEqualsOpenMesh',
      matchedPortPath,
      timestampMs: Date.now(),
      ...(authHeaderPresentForDelegate !== undefined
        ? { authHeaderPresent: authHeaderPresentForDelegate }
        : {}),
    });
    return await innerFetch(req, env, ctx);
  };
}

// ─── describeMeshPolicyPosture ───────────────────────────────────────

export interface PostureReport {
  readonly totalPorts: number;
  readonly overlayCount: number;
  readonly netReachablePortCount: number;
  readonly authRequired: boolean;
  readonly reachablePorts: readonly string[];
  readonly blockedPaths: readonly string[];
}

interface PostureSummary {
  readonly totalPorts: number;
  readonly overlayCount: number;
  readonly netReachablePortCount: number;
  readonly authRequired: boolean;
}

function describeCore(policy: MeshPolicyConfig): PostureSummary {
  const totalPorts = policy.ports.length;
  const overlayCount = policy.overlays?.length ?? 0;
  const authRequired =
    (policy.requireAuthHeaderPresent ?? policy.requireAuth) === true;

  const size = totalPorts;
  const portCount = totalPorts;

  if (size === 0) {
    return { totalPorts, overlayCount, netReachablePortCount: 0, authRequired };
  }

  // An overlay contributes to `overlaySize` only if it matches at least
  // one whitelisted port path, per `pathMatchesPrefix` semantics.
  let overlaySize = 0;
  if (policy.overlays) {
    for (const overlay of policy.overlays) {
      const matches = policy.ports.some((port) =>
        pathMatchesPrefix(port.path, overlay.pathPrefix),
      );
      if (matches) overlaySize += 1;
    }
  }

  if (overlaySize > portCount) {
    return { totalPorts, overlayCount, netReachablePortCount: 0, authRequired };
  }

  const mesh: Mesh = createMesh(size, portCount);
  const overlay: PortOverlay = createPortOverlay(overlaySize);
  const netReachablePortCount = meshNetHitCountWithOverlay(mesh, overlay);

  return { totalPorts, overlayCount, netReachablePortCount, authRequired };
}

/**
 * Describe the static mesh-policy posture, augmented with the lists of
 * reachable ports (not covered by any overlay) and declared overlay
 * prefixes. Result shape matches x-ranger's `posturePolicy`.
 */
export function describeMeshPolicyPosture(
  policy: MeshPolicyConfig | undefined,
): PostureReport {
  if (policy === undefined) {
    return {
      totalPorts: 0,
      overlayCount: 0,
      netReachablePortCount: 0,
      authRequired: false,
      reachablePorts: [],
      blockedPaths: [],
    };
  }

  const summary = describeCore(policy);
  const overlays = policy.overlays ?? [];

  const reachablePorts: string[] = [];
  for (const port of policy.ports) {
    const covered = overlays.some((overlay) =>
      pathMatchesPrefix(port.path, overlay.pathPrefix),
    );
    if (!covered) {
      reachablePorts.push(port.path);
    }
  }

  const blockedPaths = overlays.map((overlay) => overlay.pathPrefix);

  return {
    totalPorts: summary.totalPorts,
    overlayCount: summary.overlayCount,
    netReachablePortCount: summary.netReachablePortCount,
    authRequired: summary.authRequired,
    reachablePorts,
    blockedPaths,
  };
}

// ─── checkRequest ────────────────────────────────────────────────────

export type CheckGate = 'overlay' | 'path' | 'method' | 'auth' | 'delegate';

export interface CheckResult {
  readonly path: string;
  readonly method: string;
  readonly authHeaderPresent: boolean;
  readonly gate: CheckGate;
  readonly status: number;
  readonly reason: string;
  readonly theorem: string;
}

/**
 * Dry-run the gate chain against `(policy, path, method, authHeaderPresent)`
 * without constructing a `Request`. Mirrors the ordering of
 * {@link wrapWithMeshPolicy} exactly.
 *
 * Gate chain:
 *   1. Overlay cancellation (403, `portOverlayCancelsWithinPorts`).
 *   2. Default-deny whitelist (404, `emptyPortListSealsRegion` when
 *      ports is empty, otherwise `deploymentShapesToContainer`).
 *   3. Method filter on matched port (405).
 *   4. Auth-header-presence gate (401, `authenticationRestrictsPorts`).
 *   5. Delegate (200, `fullPortListEqualsOpenMesh`).
 */
export function checkRequest(
  policy: MeshPolicyConfig | undefined,
  path: string,
  method: string = 'GET',
  authHeaderPresent: boolean = false,
): CheckResult {
  const normalizedMethod = method.toUpperCase();

  if (policy === undefined) {
    return {
      path,
      method: normalizedMethod,
      authHeaderPresent,
      gate: 'delegate',
      status: 200,
      reason: 'no mesh policy configured',
      theorem: 'N/A',
    };
  }

  // Gate 1.
  const overlays = policy.overlays ?? [];
  for (const overlay of overlays) {
    if (pathMatchesPrefix(path, overlay.pathPrefix)) {
      return {
        path,
        method: normalizedMethod,
        authHeaderPresent,
        gate: 'overlay',
        status: 403,
        reason: `overlay-blocked: ${overlay.reason ?? 'unspecified'}`,
        theorem: 'portOverlayCancelsWithinPorts',
      };
    }
  }

  // Gate 2.
  let matchedPort: MeshPortSpec | undefined;
  for (const port of policy.ports) {
    if (pathMatchesPrefix(path, port.path)) {
      matchedPort = port;
      break;
    }
  }
  if (!matchedPort) {
    return {
      path,
      method: normalizedMethod,
      authHeaderPresent,
      gate: 'path',
      status: 404,
      reason: 'path not on port whitelist',
      theorem:
        policy.ports.length === 0
          ? 'emptyPortListSealsRegion'
          : 'deploymentShapesToContainer',
    };
  }

  // Gate 3.
  if (matchedPort.methods && matchedPort.methods.length > 0) {
    const allowed = matchedPort.methods.some(
      (m) => m.toUpperCase() === normalizedMethod,
    );
    if (!allowed) {
      return {
        path,
        method: normalizedMethod,
        authHeaderPresent,
        gate: 'method',
        status: 405,
        reason: `method not allowed on port '${matchedPort.path}'`,
        theorem: 'portOverlayCancelsWithinPorts',
      };
    }
  }

  // Gate 4.
  const requireAuthHeader =
    policy.requireAuthHeaderPresent ?? policy.requireAuth;
  if (requireAuthHeader === true && !authHeaderPresent) {
    return {
      path,
      method: normalizedMethod,
      authHeaderPresent,
      gate: 'auth',
      status: 401,
      reason: 'authentication required',
      theorem: 'authenticationRestrictsPorts',
    };
  }

  // Gate 5.
  return {
    path,
    method: normalizedMethod,
    authHeaderPresent,
    gate: 'delegate',
    status: 200,
    reason: 'all gates passed',
    theorem: 'fullPortListEqualsOpenMesh',
  };
}

// ─── auditPolicy ─────────────────────────────────────────────────────

export interface AuditFinding {
  readonly severity: 'error' | 'warning' | 'info';
  readonly code: string;
  readonly message: string;
  readonly subject?: string;
}

export interface AuditBypassReductions {
  /** Reclaimed hit count from `overlayHoleBypass`. */
  readonly overlayHoleReclaim: number;
  /** Frame deficit over horizon `H = 1000`. Cites
   *  `clockRateSideChannelUnbounded`. */
  readonly clockSideChannelFrameDeficit: number;
  /** Zenonic barrier is never reached for any finite k. Cites
   *  `zenonicApproachNumeratorFixed`. */
  readonly zenonicStepsUntilUnderflow: number | 'infinite';
}

export interface AuditReport {
  readonly policyPresent: boolean;
  readonly findings: readonly AuditFinding[];
  readonly bypassReductions: AuditBypassReductions;
}

// eslint-disable-next-line no-control-regex
const CONTROL_CHAR_PATTERN = /[\x00-\x1F\x7F]/;

function containsPathTraversal(path: string): boolean {
  if (path === '..') return true;
  const segments = path.split('/');
  return segments.some((segment) => segment === '..');
}

function validatePortPath(
  port: MeshPortSpec,
  index: number,
  findings: AuditFinding[],
): void {
  if (port.path === '' || !port.path.startsWith('/')) {
    findings.push({
      severity: 'error',
      code: 'PORT_PATH_INVALID',
      message: `port[${index}] path must be non-empty and start with '/': '${port.path}'`,
      subject: port.path,
    });
  }
}

function validateOverlayPath(
  overlay: MeshOverlaySpec,
  index: number,
  findings: AuditFinding[],
): void {
  const prefix = overlay.pathPrefix;
  if (prefix === '' || !prefix.startsWith('/')) {
    findings.push({
      severity: 'error',
      code: 'OVERLAY_PATH_INVALID',
      message: `overlay[${index}] pathPrefix must be non-empty and start with '/': '${prefix}'`,
      subject: prefix,
    });
    return;
  }
  if (containsPathTraversal(prefix)) {
    findings.push({
      severity: 'error',
      code: 'OVERLAY_PATH_INVALID',
      message: `overlay[${index}] pathPrefix contains '..' traversal: '${prefix}'`,
      subject: prefix,
    });
  }
}

function validateOverlayReason(
  overlay: MeshOverlaySpec,
  index: number,
  findings: AuditFinding[],
): void {
  if (overlay.reason === undefined) return;
  if (CONTROL_CHAR_PATTERN.test(overlay.reason)) {
    findings.push({
      severity: 'warning',
      code: 'OVERLAY_REASON_CONTROL_CHARS',
      message: `overlay[${index}] reason contains control characters (injection risk)`,
      subject: overlay.reason,
    });
  }
}

function overlayMatchesAnyPort(
  overlay: MeshOverlaySpec,
  ports: readonly MeshPortSpec[],
): boolean {
  for (const port of ports) {
    if (pathMatchesPrefix(port.path, overlay.pathPrefix)) return true;
  }
  return false;
}

function computeBypassReductions(
  portCount: number,
  overlayCount: number,
): AuditBypassReductions {
  const regionSize = portCount + overlayCount + 1;
  const hole = overlayHoleBypass(
    createOverlay(overlayCount),
    createOverlay(1),
    regionSize,
  );
  const clock = clockSideChannelBypass(createClock(1), createClock(100), 1000);
  // Invoke the Zenonic witness to document the theorem path.
  zenonicApproachBypass(createGap(1, 1), 1000);

  return {
    overlayHoleReclaim: hole.reclaimedHits,
    clockSideChannelFrameDeficit: clock.frameDeficit,
    zenonicStepsUntilUnderflow: 'infinite',
  };
}

/**
 * Statically audit a mesh-policy config. Output shape matches
 * x-ranger's `auditPolicy` exactly so migrating consumers is a no-op.
 */
export function auditPolicy(
  policy: MeshPolicyConfig | undefined,
): AuditReport {
  const findings: AuditFinding[] = [];

  if (policy === undefined) {
    return {
      policyPresent: false,
      findings,
      bypassReductions: computeBypassReductions(0, 0),
    };
  }

  if (policy.ports.length === 0) {
    findings.push({
      severity: 'warning',
      code: 'EMPTY_PORTS',
      message:
        'policy has no ports; every path 404s (emptyPortListSealsRegion)',
    });
  }

  policy.ports.forEach((port, index) => {
    validatePortPath(port, index, findings);
  });

  const overlays = policy.overlays ?? [];
  overlays.forEach((overlay, index) => {
    validateOverlayPath(overlay, index, findings);
    validateOverlayReason(overlay, index, findings);
    const pathInvalid =
      overlay.pathPrefix === '' ||
      !overlay.pathPrefix.startsWith('/') ||
      containsPathTraversal(overlay.pathPrefix);
    if (!pathInvalid && !overlayMatchesAnyPort(overlay, policy.ports)) {
      findings.push({
        severity: 'info',
        code: 'OVERLAY_DEAD',
        message: `overlay[${index}] pathPrefix '${overlay.pathPrefix}' matches no port; it will never block (possible misconfig)`,
        subject: overlay.pathPrefix,
      });
    }
  });

  const posture = describeCore(policy);
  if (policy.ports.length > 0 && posture.netReachablePortCount === 0) {
    findings.push({
      severity: 'error',
      code: 'TOTAL_MESH_COLLAPSE',
      message:
        'The configured overlays completely cover all ports. The mesh is topologically collapsed and unreachable.',
      subject: 'mesh.overlays',
    });
  }

  const authPresence =
    policy.requireAuthHeaderPresent ?? policy.requireAuth ?? false;
  if (authPresence === true) {
    findings.push({
      severity: 'info',
      code: 'AUTH_PRESENCE_ONLY',
      message:
        'requireAuth(HeaderPresent) is true; presence-only check — no token format, signature, or expiration is validated',
    });
  }

  const totalOverlaySize = overlays.length;
  return {
    policyPresent: true,
    findings,
    bypassReductions: computeBypassReductions(
      policy.ports.length,
      totalOverlaySize,
    ),
  };
}

// ─── Linked-service helpers ──────────────────────────────────────────

/**
 * Lift a mesh-aware service binding into a formal `Link` from
 * `@a0n/gnosis/mesh-policy`. Returns `null` when the binding carries
 * no mesh metadata or when either index is out of range against its
 * respective `Mesh`'s port count. Cites `linkRequiresSourcePort` and
 * `linkRequiresDestPort`.
 */
export function buildLinkFromBinding(
  binding: MeshServiceBinding,
  sourceMesh: Mesh,
  destMesh: Mesh,
): Link | null {
  const mesh = binding.mesh;
  if (!mesh) return null;
  if (mesh.sourcePortIdx >= sourceMesh.portCount) return null;
  if (mesh.destPortIdx >= destMesh.portCount) return null;
  return createLink(mesh.sourcePortIdx, mesh.destPortIdx);
}

export interface LinkedServiceAuditFinding {
  readonly severity: 'error' | 'warning' | 'info';
  readonly code: string;
  readonly message: string;
  /** The binding name, for example `'AUTH_SERVICE'`. */
  readonly binding: string;
}

export interface LinkedServiceAuditReport {
  readonly totalServices: number;
  readonly meshAwareCount: number;
  /** Bindings without `mesh` metadata (legacy style). */
  readonly legacyCount: number;
  readonly findings: readonly LinkedServiceAuditFinding[];
}

function portCoveredByOverlay(
  port: MeshPortSpec,
  overlays: readonly MeshOverlaySpec[],
): boolean {
  for (const overlay of overlays) {
    if (pathMatchesPrefix(port.path, overlay.pathPrefix)) return true;
  }
  return false;
}

/**
 * Audit a set of service bindings against the local mesh policy.
 *
 * Findings match x-ranger's `auditLinkedServices`:
 *   - `SERVICE_NOT_MESH_AWARE` (info)
 *   - `SERVICE_WITHOUT_LOCAL_MESH` (warning)
 *   - `SERVICE_SOURCE_PORT_OUT_OF_RANGE` (error)
 *   - `SERVICE_SOURCE_PORT_COVERED_BY_OVERLAY` (warning)
 */
export function auditLinkedServices(
  bindings: readonly MeshServiceBinding[],
  localMeshPolicy: MeshPolicyConfig | undefined,
): LinkedServiceAuditReport {
  const findings: LinkedServiceAuditFinding[] = [];
  let totalServices = 0;
  let meshAwareCount = 0;
  let legacyCount = 0;

  const overlays = localMeshPolicy?.overlays ?? [];

  for (const binding of bindings) {
    totalServices += 1;

    if (!binding.mesh) {
      legacyCount += 1;
      findings.push({
        severity: 'info',
        code: 'SERVICE_NOT_MESH_AWARE',
        message: `service binding '${binding.binding}' has no mesh metadata (legacy style)`,
        binding: binding.binding,
      });
      continue;
    }

    meshAwareCount += 1;

    if (localMeshPolicy === undefined) {
      findings.push({
        severity: 'warning',
        code: 'SERVICE_WITHOUT_LOCAL_MESH',
        message: `service binding '${binding.binding}' declares mesh metadata but the local config has no [mesh] section to validate against`,
        binding: binding.binding,
      });
      continue;
    }

    const sourcePortIdx = binding.mesh.sourcePortIdx;
    const ports = localMeshPolicy.ports;
    if (sourcePortIdx >= ports.length) {
      findings.push({
        severity: 'error',
        code: 'SERVICE_SOURCE_PORT_OUT_OF_RANGE',
        message: `service binding '${binding.binding}' sourcePortIdx=${sourcePortIdx} is out of range (local mesh has ${ports.length} port(s))`,
        binding: binding.binding,
      });
      continue;
    }

    const sourcePort = ports[sourcePortIdx] as MeshPortSpec;
    if (
      binding.mesh.inheritOverlays &&
      overlays.length > 0 &&
      portCoveredByOverlay(sourcePort, overlays)
    ) {
      findings.push({
        severity: 'warning',
        code: 'SERVICE_SOURCE_PORT_COVERED_BY_OVERLAY',
        message: `service binding '${binding.binding}' source port '${sourcePort.path}' is covered by a local overlay; the link will be blocked by linkHonorsSourceOverlay`,
        binding: binding.binding,
      });
    }
  }

  return {
    totalServices,
    meshAwareCount,
    legacyCount,
    findings,
  };
}
