// @a0n/gnosis/src/mesh-engine.ts
/**
 * Mesh Engine -- Async Fork/Race/Fold on a Scheduling Mesh
 *
 * The three primitives as real concurrent operations:
 *   fork: dispatch work to k vertices simultaneously
 *   race: resolve on first completion
 *   fold: await all results
 *
 * Built on SchedulingMesh topology. God formula governs weights.
 * Pathfinding routes between cells. The mesh formalizes the scheduler.
 *
 * @packageDocumentation
 */

import { godFormula } from './holy-trinity';
import { SchedulingMesh, type MeshVertex, type MeshEvent, FORK, VOID, VENT, type Trit } from './scheduling-mesh';
import { type GnosticMeshLevel } from './spiderweb-topology';

// ─── Types ──────────────────────────────────────────────────────────

/** A unit of work that runs on a vertex. */
export type Task<T> = (vertex: MeshVertex, signal: AbortSignal) => Promise<T>;

/** Result of an async fork. */
export interface AsyncForkResult<T> {
  readonly hub: number;
  readonly promises: Array<{ vertexId: number; promise: Promise<T> }>;
  readonly abort: AbortController;
}

/** Result of an async race. */
export interface AsyncRaceResult<T> {
  readonly winner: number;
  readonly result: T;
  readonly elapsed: number;
  readonly losers: number[];
}

/** Result of an async fold. */
export interface AsyncFoldResult<T> {
  readonly hub: number;
  readonly results: Array<{ vertexId: number; result: T }>;
  readonly elapsed: number;
}

// ─── Engine ─────────────────────────────────────────────────────────

/**
 * MeshEngine -- runs async tasks on a SchedulingMesh.
 *
 * The topology routes. The engine executes. The god formula learns.
 */
export class MeshEngine {
  readonly mesh: SchedulingMesh;
  private rounds: number = 0;
  private listeners: Array<(event: MeshEvent) => void> = [];

  constructor(mesh: SchedulingMesh) {
    this.mesh = mesh;
  }

  on(listener: (event: MeshEvent) => void): void {
    this.listeners.push(listener);
  }

  private emit(event: MeshEvent): void {
    for (const fn of this.listeners) fn(event);
  }

  /**
   * P1 FORK -- dispatch a task to all k radial vertices from a hub.
   * Each vertex runs the task concurrently. Returns promises.
   */
  fork<T>(hubId: number, task: Task<T>): AsyncForkResult<T> {
    const hub = this.mesh.vertices[hubId];
    if (!hub || hub.type !== 'hub') {
      throw new Error(`Vertex ${hubId} is not a hub`);
    }

    const edges = this.mesh.adjacency.get(hubId) ?? [];
    const radials = edges.filter((e: any) => e.type === 'radial');
    const abort = new AbortController();

    hub.state = FORK;
    this.rounds++;

    const promises = radials.map((edge: any) => {
      const vertexId = edge.to;
      if (typeof vertexId !== 'number') {
        throw new Error('Radial edge is missing a target vertex');
      }
      const vertex = this.mesh.vertices[vertexId];
      if (vertex === undefined) {
        throw new Error(`Missing vertex ${vertexId} in fork`);
      }
      vertex.state = FORK;
      vertex.payload = task;

      const promise = task(vertex, abort.signal).finally(() => {
        vertex.state = VOID;
        vertex.payload = null;
      });

      return { vertexId: edge.to, promise };
    });

    this.emit({ type: 'fork', hub: hubId, targets: promises.map((p: any) => p.vertexId), hop: this.rounds });

    return { hub: hubId, promises, abort };
  }

  /**
   * P3 RACE -- first vertex to complete wins. Losers are aborted.
   * God formula updates: losers get rejected, winner keeps weight.
   */
  async race<T>(forked: AsyncForkResult<T>): Promise<AsyncRaceResult<T>> {
    const start = Date.now();

    // Race: Promise.race but we need to know WHICH one won
    const tagged = forked.promises.map(({ vertexId, promise }) =>
      promise.then(result => ({ vertexId, result }))
    );

    const winner = await Promise.race(tagged);

    // Abort the losers
    forked.abort.abort();

    const losers: number[] = [];
    for (const { vertexId } of forked.promises) {
      const vertex = this.mesh.vertices[vertexId];
      if (vertex === undefined) {
        continue;
      }
      if (vertexId === winner.vertexId) {
        // Winner: keep weight
        vertex.busy = false;
      } else {
        // Loser: reject
        vertex.rejections++;
        vertex.weight = godFormula(this.rounds, vertex.rejections);
        vertex.busy = false;
        losers.push(vertexId);
      }
    }

    // Release hub
    const hubVertex = this.mesh.vertices[forked.hub];
    if (hubVertex === undefined) {
      throw new Error(`Missing hub vertex ${forked.hub}`);
    }
    hubVertex.busy = false;

    this.emit({ type: 'race', winner: winner.vertexId, losers, hop: this.rounds });

    return {
      winner: winner.vertexId,
      result: winner.result,
      elapsed: Date.now() - start,
      losers,
    };
  }

  /**
   * P2 FOLD -- await ALL results. No abort, no rejection.
   * Every vertex contributes. Collect everything.
   */
  async fold<T>(forked: AsyncForkResult<T>): Promise<AsyncFoldResult<T>> {
    const start = Date.now();

    const settled = await Promise.allSettled(
      forked.promises.map(({ vertexId, promise }) =>
        promise.then(result => ({ vertexId, result }))
      )
    );

    const results: Array<{ vertexId: number; result: T }> = [];
    for (const s of settled) {
      if (s.status === 'fulfilled') {
        results.push(s.value);
      }
    }

    // Release hub
    const hubVertex = this.mesh.vertices[forked.hub];
    if (hubVertex === undefined) {
      throw new Error(`Missing hub vertex ${forked.hub}`);
    }
    hubVertex.busy = false;

    this.emit({ type: 'fold', hub: forked.hub, hop: this.rounds });

    return {
      hub: forked.hub,
      results,
      elapsed: Date.now() - start,
    };
  }

  /**
   * Fork → Race: the most common pattern.
   * Dispatch to k vertices, first one wins, losers rejected.
   */
  async forkRace<T>(hubId: number, task: Task<T>): Promise<AsyncRaceResult<T>> {
    const forked = this.fork(hubId, task);
    return this.race(forked);
  }

  /**
   * Fork → Fold: dispatch to k vertices, collect all results.
   */
  async forkFold<T>(hubId: number, task: Task<T>): Promise<AsyncFoldResult<T>> {
    const forked = this.fork(hubId, task);
    return this.fold(forked);
  }

  /**
   * Fork → Race → Fold: the full cascade.
   *
   * 1. Fork from hub along k radials
   * 2. Race: first result wins
   * 3. Pathfind to nearest idle hub
   * 4. Fork again from that hub
   * 5. Repeat until no idle hubs or maxHops
   */
  async cascade<T>(
    startHub: number,
    task: Task<T>,
    maxHops: number = this.mesh.cellCount,
  ): Promise<T[]> {
    const results: T[] = [];
    let currentHub = startHub;

    for (let hop = 0; hop < maxHops; hop++) {
      const raceResult = await this.forkRace(currentHub, task);
      results.push(raceResult.result);

      // Find nearest idle hub
      const idleHubs = this.mesh.idleHubs();
      if (idleHubs.length === 0) {
        this.emit({ type: 'saturated', hop });
        break;
      }

      let nearestHub: number | null = null;
      let nearestDist = Infinity;
      for (const candidate of idleHubs) {
        const path = this.mesh.shortestPath(raceResult.winner, candidate);
        if (path && path.length < nearestDist) {
          nearestDist = path.length;
          nearestHub = candidate;
        }
      }

      if (nearestHub === null) break;

      this.emit({ type: 'propagate', from: raceResult.winner, to: nearestHub, hops: nearestDist, hop });
      currentHub = nearestHub;
    }

    return results;
  }

  /**
   * Fire all hubs in parallel -- full mesh activation.
   * Every hub forks simultaneously. All races run concurrently.
   */
  async fireAll<T>(task: Task<T>): Promise<AsyncRaceResult<T>[]> {
    const idleHubs = this.mesh.idleHubs();
    const racePromises = idleHubs.map(hubId => this.forkRace(hubId, task));
    return Promise.all(racePromises);
  }

  /**
   * Route -- pathfind from any vertex to any vertex, running a task
   * at each hop along the shortest path. Each vertex transforms
   * the payload before passing it to the next.
   *
   * This is the fundamental mesh operation: a signal propagating
   * through the topology, transformed at each vertex it touches.
   */
  async route<T>(
    from: number,
    to: number,
    initial: T,
    transform: (vertex: MeshVertex, payload: T, hop: number, signal: AbortSignal) => Promise<T>,
  ): Promise<{ path: number[]; result: T; elapsed: number } | null> {
    const path = this.mesh.shortestPath(from, to);
    if (!path) return null;

    const start = Date.now();
    const abort = new AbortController();
    let payload = initial;

    for (let i = 0; i < path.length; i++) {
      const vertexId = path[i];
      if (vertexId === undefined) {
        throw new Error('Walk path is missing a vertex id');
      }
      const vertex = this.mesh.vertices[vertexId];
      if (vertex === undefined) {
        throw new Error(`Missing vertex ${vertexId} in walk`);
      }
      vertex.state = FORK;
      payload = await transform(vertex, payload, i, abort.signal);
      vertex.state = VOID;
    }

    return { path, result: payload, elapsed: Date.now() - start };
  }

  /**
   * Scatter -- send a task outward from a source vertex to all
   * vertices within a hop radius. Like fork but not limited to
   * one hub's radials. Reaches across cell boundaries.
   */
  async scatter<T>(
    source: number,
    radius: number,
    task: Task<T>,
  ): Promise<Array<{ vertexId: number; result: T; hops: number }>> {
    const results: Array<{ vertexId: number; result: T; hops: number }> = [];
    const visited = new Set<number>();
    const queue: Array<{ id: number; hops: number }> = [{ id: source, hops: 0 }];
    const promises: Array<{ vertexId: number; hops: number; promise: Promise<T> }> = [];
    const abort = new AbortController();

    visited.add(source);

    while (queue.length > 0) {
      const { id, hops } = queue.shift()!;
      if (hops > radius) continue;

      const vertex = this.mesh.vertices[id];
      if (vertex === undefined) {
        continue;
      }
      if (vertex.state === VOID) {
        vertex.state = FORK;
        promises.push({
          vertexId: id,
          hops,
          promise: task(vertex, abort.signal).finally(() => { vertex.state = VOID; }),
        });
      }

      if (hops < radius) {
        const neighbors = this.mesh.neighbors(id);
        for (const n of neighbors) {
          if (!visited.has(n)) {
            visited.add(n);
            queue.push({ id: n, hops: hops + 1 });
          }
        }
      }
    }

    const settled = await Promise.allSettled(
      promises.map(p => p.promise.then(result => ({ vertexId: p.vertexId, result, hops: p.hops })))
    );

    for (const s of settled) {
      if (s.status === 'fulfilled') results.push(s.value);
    }

    return results;
  }

  /**
   * Gather -- collect results from all vertices within a hop radius
   * back to a destination vertex. The inverse of scatter.
   */
  async gather<T, R>(
    destination: number,
    radius: number,
    extract: Task<T>,
    reduce: (accumulated: R, next: T, vertexId: number) => R,
    initial: R,
  ): Promise<{ result: R; sources: number; elapsed: number }> {
    const start = Date.now();
    const scattered = await this.scatter(destination, radius, extract);
    let result = initial;
    for (const s of scattered) {
      result = reduce(result, s.result, s.vertexId);
    }
    return { result, sources: scattered.length, elapsed: Date.now() - start };
  }

  /**
   * Wormhole -- direct jump from any vertex to any other vertex.
   * Skips the path. O(1) hop. The payload teleports.
   *
   * Cost: the source and destination must both exist and the
   * destination must be idle. No intermediate vertices activate.
   * The topology is bypassed -- use sparingly.
   */
  async wormhole<T>(
    from: number,
    to: number,
    task: Task<T>,
  ): Promise<{ result: T; elapsed: number }> {
    const source = this.mesh.vertices[from];
    const dest = this.mesh.vertices[to];
    if (!dest || dest.state !== VOID) throw new Error(`Destination ${to} busy or missing`);

    const start = Date.now();
    const abort = new AbortController();
    dest.state = FORK;

    const result = await task(dest, abort.signal);
    dest.state = VOID;

    return { result, elapsed: Date.now() - start };
  }

  /**
   * Bidirectional search -- burn the candle from both ends.
   *
   * Wormhole to the destination, then expand BFS from BOTH source
   * and destination simultaneously. When the two wavefronts meet,
   * the shortest path is found in O(√n) instead of O(n).
   *
   * Returns the meeting point and the full path.
   */
  bidirectionalFind(from: number, to: number): { path: number[]; meetAt: number } | null {
    if (from === to) return { path: [from], meetAt: from };

    const forwardVisited = new Map<number, number>(); // vertex → parent
    const backwardVisited = new Map<number, number>();
    const forwardQueue: number[] = [from];
    const backwardQueue: number[] = [to];

    forwardVisited.set(from, -1);
    backwardVisited.set(to, -1);

    while (forwardQueue.length > 0 || backwardQueue.length > 0) {
      // Expand forward one step
      if (forwardQueue.length > 0) {
        const current = forwardQueue.shift()!;
        for (const n of this.mesh.neighbors(current)) {
          if (!forwardVisited.has(n)) {
            forwardVisited.set(n, current);
            if (backwardVisited.has(n)) {
              return this.reconstructBiPath(n, forwardVisited, backwardVisited);
            }
            forwardQueue.push(n);
          }
        }
      }

      // Expand backward one step
      if (backwardQueue.length > 0) {
        const current = backwardQueue.shift()!;
        for (const n of this.mesh.neighbors(current)) {
          if (!backwardVisited.has(n)) {
            backwardVisited.set(n, current);
            if (forwardVisited.has(n)) {
              return this.reconstructBiPath(n, forwardVisited, backwardVisited);
            }
            backwardQueue.push(n);
          }
        }
      }
    }

    return null;
  }

  private reconstructBiPath(
    meetAt: number,
    forwardVisited: Map<number, number>,
    backwardVisited: Map<number, number>,
  ): { path: number[]; meetAt: number } {
    // Forward path: from → meetAt
    const forwardPath: number[] = [];
    let node = meetAt;
    while (node !== -1) {
      forwardPath.unshift(node);
      node = forwardVisited.get(node) ?? -1;
    }

    // Backward path: meetAt → to
    node = backwardVisited.get(meetAt) ?? -1;
    while (node !== -1) {
      forwardPath.push(node);
      node = backwardVisited.get(node) ?? -1;
    }

    return { path: forwardPath, meetAt };
  }

  /**
   * Wormhole-assisted route -- jump to destination, then search
   * from both ends to find and activate the optimal path.
   *
   * 1. Wormhole: plant a beacon at the destination
   * 2. Bidirectional BFS: find shortest path from both ends
   * 3. Route: run the transform along the found path
   *
   * Faster than route alone for distant vertices in large meshes.
   */
  async wormholeRoute<T>(
    from: number,
    to: number,
    initial: T,
    transform: (vertex: MeshVertex, payload: T, hop: number, signal: AbortSignal) => Promise<T>,
  ): Promise<{ path: number[]; meetAt: number; result: T; elapsed: number } | null> {
    const found = this.bidirectionalFind(from, to);
    if (!found) return null;

    const start = Date.now();
    const abort = new AbortController();
    let payload = initial;

    for (let i = 0; i < found.path.length; i++) {
      const vertexId = found.path[i];
      if (vertexId === undefined) {
        throw new Error('Bidirectional path is missing a vertex id');
      }
      const vertex = this.mesh.vertices[vertexId];
      if (vertex === undefined) {
        throw new Error(`Missing vertex ${vertexId} in bidirectional walk`);
      }
      vertex.state = FORK;
      payload = await transform(vertex, payload, i, abort.signal);
      vertex.state = VOID;
    }

    return { ...found, result: payload, elapsed: Date.now() - start };
  }

  /**
   * Delayed Cascade -- fork WITHOUT folding for N hops, then project once.
   *
   * Classical cascade: fork → race → fork → race → ... (projects at every hop)
   * Delayed cascade:   fork → fork → fork → ... → race (projects once at end)
   *
   * The delayed approach preserves 3^D states vs 2^D states for D hops.
   * The information advantage grows exponentially with delay depth.
   *
   * From ShannonMirroredRooms.lean:
   *   speedup_is_projection_delay: quantum speedup is the dividend of
   *   deferring the projection (measurement) to the end of the pipeline.
   *
   * @param startHub - Hub to start from
   * @param task - Task to fork to each vertex
   * @param delayHops - Number of hops to delay projection (default: √N)
   * @returns Results from the single projection, plus delay metrics
   */
  async delayedCascade<T>(
    startHub: number,
    task: Task<T>,
    delayHops?: number,
  ): Promise<{
    results: T[];
    delayDepth: number;
    tritStates: number;
    bitStates: number;
    advantageRatio: number;
    elapsed: number;
  }> {
    const start = Date.now();
    const N = this.mesh.hubs.length;
    const maxDelay = delayHops ?? Math.max(1, Math.floor(Math.sqrt(N)));

    // Phase 1: Delayed forks — stay in trit space
    const allForked: AsyncForkResult<T>[] = [];
    let currentHub = startHub;

    for (let hop = 0; hop < maxDelay; hop++) {
      const forked = this.fork(currentHub, task);
      allForked.push(forked);

      this.emit({ type: 'delayedFork', hub: currentHub, hop, depth: hop + 1 });

      // Find next idle hub (but don't race yet)
      const idleHubs = this.mesh.idleHubs();
      if (idleHubs.length === 0) break;

      let nearestHub: number | null = null;
      let nearestDist = Infinity;
      for (const candidate of idleHubs) {
        if (candidate === currentHub) continue;
        const path = this.mesh.shortestPath(currentHub, candidate);
        if (path && path.length < nearestDist) {
          nearestDist = path.length;
          nearestHub = candidate;
        }
      }

      if (nearestHub === null) break;
      currentHub = nearestHub;
    }

    const delayDepth = allForked.length;

    // Phase 2: Single projection — race all accumulated forks
    const results: T[] = [];
    for (const forked of allForked) {
      const raceResult = await this.race(forked);
      results.push(raceResult.result);
    }

    const tritStates = 3 ** delayDepth;
    const bitStates = 2 ** delayDepth;

    this.emit({ type: 'projection', depth: delayDepth, tritStates, bitStates });

    return {
      results,
      delayDepth,
      tritStates,
      bitStates,
      advantageRatio: delayDepth === 0 ? 1 : tritStates / bitStates,
      elapsed: Date.now() - start,
    };
  }
}

// ─── Factory ──────────────────────────────────────────────────────

/**
 * Create a mesh engine at a gnostic level.
 *
 * @example
 * const engine = createEngine('proton', 3);
 * const result = await engine.forkRace(0, async (vertex, signal) => {
 *   // do work
 *   return vertex.id * 2;
 * });
 * console.log(result.winner, result.result);
 */
export function createEngine(level: GnosticMeshLevel, cells: number): MeshEngine {
  return new MeshEngine(new SchedulingMesh(level, cells));
}
