// @a0n/gnosis/src/mesh-policy-compile.ts
/**
 * Mesh-policy compile pipeline — TypeScript mirror of the Lean
 * `compileMeshPolicy` function from
 * `MeshPolicyCompileCorrectness.lean`.
 *
 * Serializes a {@link MeshPolicyConfig} from `./mesh-policy-edge.ts`
 * into a `.bw` binary container, and exposes the inverse load
 * operation that reconstructs the policy byte-for-byte.
 *
 * The container format (kind=2 — mesh-policy) is a length-prefixed
 * header followed by the fp48-packed payload:
 *
 *     magic     : 4 bytes   "BZRO"
 *     version   : 1 byte    current = 1
 *     kind      : 1 byte    2 = mesh-policy (self-describing)
 *     manifLen  : 4 bytes   u32 little-endian manifest JSON length
 *     manifest  : manifLen  UTF-8 JSON bytes of the compact manifest
 *                           (carries the opaque-pack fields needed to
 *                           reverse the payload byte layout)
 *     payload   : tail      `packOpaqueBytes(policyJson)` -- fp48
 *                           packed low-48 surface + parity sidecar +
 *                           verbatim tail, concatenated
 *
 * The payload is produced by `@a0n/bitwise/opaque-bytes`'s
 * `packOpaqueBytes`, which applies the fp48 byte-packing primitive to
 * arbitrary `Uint8Array` payloads. The compressed-surface length is
 * bounded by `ceil(policyJsonLength * 3/4)` per Lean
 * `BizarroCompiler.bitwise_shrink_4_3`; parity + tail are carried
 * inline for lossless round-trip.
 *
 * Determinism is load-bearing: callers that want reproducible builds
 * pass a fixed `options.timestamp`. The default is `0` (epoch), so
 * tests see byte-identical output across runs. Real deploys inject
 * `Date.now()` explicitly.
 *
 * No Node-specific imports. Uses Web APIs only (`TextEncoder`,
 * `TextDecoder`, `DataView`, `Uint8Array`).
 */

import {
  opaqueBytesShrinkBound,
  packOpaqueBytes,
  unpackOpaqueBytes,
  type OpaquePackManifest,
} from '@a0n/bitwise/opaque-bytes';
import type {
  AdoptionMode,
  MeshOverlaySpec,
  MeshPolicyConfig,
  MeshPortSpec,
} from './mesh-policy-edge';

// ─── Container framing constants ─────────────────────────────────────

/** Magic header bytes: ASCII "BZRO". Re-uses the bitwise magic to keep
 *  `.bw` files visually and structurally cohesive across kinds. */
const MAGIC_BYTES = new Uint8Array([0x42, 0x5a, 0x52, 0x4f]); // "BZRO"

/** Container version byte. Bump on breaking framing change. */
const CONTAINER_VERSION = 1;

/** Kind byte for mesh-policy-specific containers. Chosen = 2 so the
 *  value is self-describing and does not collide with bitwise's
 *  existing opaque (kind=1) path. */
const KIND_MESH_POLICY = 2;

/** Manifest schema sentinel. Bump when the manifest shape changes in
 *  a backwards-incompatible way so consumers can reject old bundles. */
const MANIFEST_FORMAT = 'mesh-policy-v1' as const;

/** Fixed offsets into the container header. */
const OFFSET_MAGIC = 0;
const OFFSET_VERSION = 4;
const OFFSET_KIND = 5;
const OFFSET_MANIFEST_LEN = 6;
const OFFSET_MANIFEST_BODY = 10;

// ─── Public types ────────────────────────────────────────────────────

export interface MeshPolicyCompiledManifest {
  /** Format sentinel — must equal {@link MANIFEST_FORMAT}. */
  readonly format: typeof MANIFEST_FORMAT;
  /** Mirrors Lean `compileMeshPolicy.portCount`. */
  readonly portCount: number;
  /** Mirrors Lean `compileMeshPolicy.overlayCount`. */
  readonly overlayCount: number;
  /** Mirrors Lean `compileMeshPolicy.authFlag`. */
  readonly authFlag: boolean;
  /** Lean `modeCode`: 0 = enforce, 1 = audit, 2 = shadow. */
  readonly modeCode: number;
  /** Whether any overlay carries the honeypit flag. */
  readonly honeypitPresent: boolean;
  /** Build-audit ISO timestamp. Deterministic-by-default so tests
   *  compare byte-identical across runs. Real deploys override. */
  readonly compiledAt: string;
  /** Bytes of the unpacked policy JSON (pre-framing). Mirrors
   *  `OpaquePackManifest.originalLength`. */
  readonly nominalSize: number;
  /** Bytes of the final `.bw` container. */
  readonly compiledSize: number;
  /**
   * Size of the compressed fp48 surface inside the payload section.
   * Bounded by `ceil(nominalSize * 3/4)` per
   * `BizarroCompiler.bitwise_shrink_4_3`.
   */
  readonly packedLength: number;
  /**
   * Size of the fp48 parity sidecar inside the payload section.
   * Always `2 * wordCount` for opaque-bytes' 8-byte word quantum.
   */
  readonly parityBytes: number;
  /**
   * Trailing unpacked bytes inside the payload section (range
   * `[0, 8)`).
   */
  readonly tailBytes: number;
  /**
   * SHA-256 content-address of the compiled `.bw` bundle, rendered as
   * 64 lowercase hex chars. Enables idempotent deploys: compute this
   * locally, compare against the deployed fingerprint, skip re-shipping
   * when equal. Anchored by the Lean `compile_deterministic` theorem —
   * equal input policies produce equal bundles, hence equal fingerprints.
   *
   * Populated synchronously by {@link compileMeshPolicy} via a pure-JS
   * SHA-256 (no Web Crypto / Node `crypto` / `Buffer` dependency).
   */
  readonly bundleFingerprint: string;
}

export interface MeshPolicyCompiledBundle {
  /** The `.bw`-formatted binary artifact. */
  readonly bwBytes: Uint8Array;
  /** Compact manifest extracted from the bundle, for diagnostics. */
  readonly manifest: MeshPolicyCompiledManifest;
}

export interface CompileOptions {
  /** Overrides the compiledAt timestamp. Default: `'1970-01-01T00:00:00.000Z'`
   *  for determinism. Pass `new Date().toISOString()` in real deploys. */
  readonly timestamp?: string;
}

// ─── JSON stable stringify ───────────────────────────────────────────

/** Sort object keys alphabetically at every depth. Strips `undefined`
 *  values (matching `JSON.stringify` behavior) so two configs that
 *  differ only in key order produce byte-identical JSON. Arrays keep
 *  their original order — they're semantically meaningful here. */
function stableStringify(value: unknown): string {
  return JSON.stringify(sortKeysDeep(value));
}

function sortKeysDeep(value: unknown): unknown {
  if (value === null || typeof value !== 'object') return value;
  if (Array.isArray(value)) return value.map(sortKeysDeep);
  const entries = Object.keys(value as Record<string, unknown>)
    .filter((k) => (value as Record<string, unknown>)[k] !== undefined)
    .sort();
  const out: Record<string, unknown> = {};
  for (const key of entries) {
    out[key] = sortKeysDeep((value as Record<string, unknown>)[key]);
  }
  return out;
}

// ─── Manifest helpers ────────────────────────────────────────────────

function adoptionModeToCode(mode: AdoptionMode | undefined): number {
  if (mode === 'audit') return 1;
  if (mode === 'shadow') return 2;
  return 0;
}

function computeAuthFlag(policy: MeshPolicyConfig): boolean {
  return (policy.requireAuthHeaderPresent ?? policy.requireAuth) === true;
}

function computeHoneypitPresent(policy: MeshPolicyConfig): boolean {
  const overlays = policy.overlays ?? [];
  for (const overlay of overlays) {
    if (overlay.honeypit === true) return true;
  }
  return false;
}

// ─── Binary framing ──────────────────────────────────────────────────

function writeU32LE(target: Uint8Array, offset: number, value: number): void {
  const view = new DataView(
    target.buffer,
    target.byteOffset,
    target.byteLength,
  );
  view.setUint32(offset, value >>> 0, true);
}

function readU32LE(source: Uint8Array, offset: number): number {
  const view = new DataView(
    source.buffer,
    source.byteOffset,
    source.byteLength,
  );
  return view.getUint32(offset, true);
}

function verifyContainerHeader(bwBytes: Uint8Array): void {
  if (bwBytes.length < OFFSET_MANIFEST_BODY) {
    throw new RangeError(
      `mesh-policy-compile: bundle too short (${bwBytes.length} < ${OFFSET_MANIFEST_BODY})`,
    );
  }
  for (let i = 0; i < 4; i++) {
    if (bwBytes[OFFSET_MAGIC + i] !== MAGIC_BYTES[i]) {
      throw new Error(
        `mesh-policy-compile: magic mismatch at byte ${i} (expected 'BZRO')`,
      );
    }
  }
  const version = bwBytes[OFFSET_VERSION];
  if (version !== CONTAINER_VERSION) {
    throw new Error(
      `mesh-policy-compile: unsupported container version ${version} (expected ${CONTAINER_VERSION})`,
    );
  }
  const kind = bwBytes[OFFSET_KIND];
  if (kind !== KIND_MESH_POLICY) {
    throw new Error(
      `mesh-policy-compile: kind ${kind} is not mesh-policy (expected ${KIND_MESH_POLICY})`,
    );
  }
}

// ─── Serialize / deserialize policy payload ──────────────────────────

interface SerializedPort {
  readonly path: string;
  readonly methods?: readonly string[];
}

interface SerializedOverlay {
  readonly pathPrefix: string;
  readonly reason?: string;
  readonly honeypit?: boolean;
}

interface SerializedPolicyPayload {
  readonly version?: number;
  readonly ports: readonly SerializedPort[];
  readonly requireAuth?: boolean;
  readonly requireAuthHeaderPresent?: boolean;
  readonly overlays?: readonly SerializedOverlay[];
  readonly honeypit?: { readonly capacity: number; readonly drainSlots: number };
  readonly mode?: AdoptionMode;
}

function policyToPayload(policy: MeshPolicyConfig): SerializedPolicyPayload {
  const ports: SerializedPort[] = policy.ports.map((p: MeshPortSpec) => {
    const out: SerializedPort =
      p.methods && p.methods.length > 0
        ? { path: p.path, methods: [...p.methods] }
        : { path: p.path };
    return out;
  });
  const overlays: SerializedOverlay[] | undefined = policy.overlays
    ? policy.overlays.map((o: MeshOverlaySpec) => {
        const out: SerializedOverlay = { pathPrefix: o.pathPrefix };
        if (o.reason !== undefined) {
          (out as { -readonly [K in keyof SerializedOverlay]: SerializedOverlay[K] }).reason = o.reason;
        }
        if (o.honeypit !== undefined) {
          (out as { -readonly [K in keyof SerializedOverlay]: SerializedOverlay[K] }).honeypit = o.honeypit;
        }
        return out;
      })
    : undefined;
  const payload: {
    -readonly [K in keyof SerializedPolicyPayload]: SerializedPolicyPayload[K];
  } = { ports };
  if (policy.version !== undefined) payload.version = policy.version;
  if (policy.requireAuth !== undefined) payload.requireAuth = policy.requireAuth;
  if (policy.requireAuthHeaderPresent !== undefined) {
    payload.requireAuthHeaderPresent = policy.requireAuthHeaderPresent;
  }
  if (overlays !== undefined) payload.overlays = overlays;
  if (policy.honeypit !== undefined) payload.honeypit = policy.honeypit;
  if (policy.mode !== undefined) payload.mode = policy.mode;
  return payload;
}

function payloadToPolicy(payload: SerializedPolicyPayload): MeshPolicyConfig {
  // Normalize `methods` back to a plain readonly array and drop the
  // key entirely when absent — this keeps round-trip equality tight
  // against the original config shape.
  const ports: MeshPortSpec[] = payload.ports.map((p) => {
    if (p.methods && p.methods.length > 0) {
      return { path: p.path, methods: [...p.methods] };
    }
    return { path: p.path };
  });
  const overlays: MeshOverlaySpec[] | undefined = payload.overlays
    ? payload.overlays.map((o) => {
        const out: {
          -readonly [K in keyof MeshOverlaySpec]: MeshOverlaySpec[K];
        } = { pathPrefix: o.pathPrefix };
        if (o.reason !== undefined) out.reason = o.reason;
        if (o.honeypit !== undefined) out.honeypit = o.honeypit;
        return out;
      })
    : undefined;
  const out: {
    -readonly [K in keyof MeshPolicyConfig]: MeshPolicyConfig[K];
  } = { ports };
  if (payload.version !== undefined) out.version = payload.version;
  if (payload.requireAuth !== undefined) out.requireAuth = payload.requireAuth;
  if (payload.requireAuthHeaderPresent !== undefined) {
    out.requireAuthHeaderPresent = payload.requireAuthHeaderPresent;
  }
  if (overlays !== undefined) out.overlays = overlays;
  if (payload.honeypit !== undefined) out.honeypit = payload.honeypit;
  if (payload.mode !== undefined) out.mode = payload.mode;
  return out;
}

// ─── Compile ─────────────────────────────────────────────────────────

/**
 * Compile a {@link MeshPolicyConfig} into a `.bw` bundle. Mirrors the
 * Lean `compileMeshPolicy` function field-by-field.
 *
 * The output is byte-deterministic for a given `(config, timestamp)`
 * pair. Stable key ordering is enforced across both the manifest and
 * the payload JSON streams.
 */
export function compileMeshPolicy(
  config: MeshPolicyConfig,
  options?: CompileOptions,
): MeshPolicyCompiledBundle {
  const encoder = new TextEncoder();

  const payload = policyToPayload(config);
  const payloadJson = stableStringify(payload);
  const payloadBytes = encoder.encode(payloadJson);

  // fp48 byte-pack the JSON payload. The packed stream is
  // self-contained (low-48 surface + parity sidecar + verbatim tail).
  // Round-trip is lossless -- `unpackOpaqueBytes(packed, opaqueManifest)`
  // returns the exact payloadBytes.
  const packResult = packOpaqueBytes(payloadBytes);
  const packedPayloadBytes = packResult.packed;
  const opaqueManifest = packResult.manifest;

  // Sanity: the compressed surface must not exceed the theoretical
  // `bitwise_shrink_4_3` bound. Guards against accidental regressions
  // in the upstream primitive.
  const shrinkBound = opaqueBytesShrinkBound(payloadBytes.length);
  if (opaqueManifest.packedLength > shrinkBound) {
    throw new Error(
      `mesh-policy-compile: packedLength ${opaqueManifest.packedLength} ` +
        `exceeds bitwise_shrink_4_3 bound ${shrinkBound}`,
    );
  }

  const timestamp = options?.timestamp ?? '1970-01-01T00:00:00.000Z';

  // Fixed-point iteration: `compiledSize` appears inside the manifest
  // JSON (so changing its digit width changes the manifest byte
  // length, which changes `compiledSize`). Converges in at most a few
  // iterations because each increase in digit width adds at most 1
  // byte; after that the byte count stabilizes. Bound iterations
  // conservatively to avoid any pathological loop.
  const baseManifest = {
    format: MANIFEST_FORMAT,
    portCount: config.ports.length,
    overlayCount: config.overlays?.length ?? 0,
    authFlag: computeAuthFlag(config),
    modeCode: adoptionModeToCode(config.mode),
    honeypitPresent: computeHoneypitPresent(config),
    compiledAt: timestamp,
    nominalSize: opaqueManifest.originalLength,
    packedLength: opaqueManifest.packedLength,
    parityBytes: opaqueManifest.parityBytes,
    tailBytes: opaqueManifest.tailBytes,
  };

  // `bundleFingerprint` is intentionally NOT part of the serialized
  // manifest JSON: the fingerprint is computed OVER the assembled
  // `bwBytes` and would otherwise form a self-reference cycle. The
  // fixed-point here converges on `compiledSize` only.
  type SerializedManifestShape = Omit<
    MeshPolicyCompiledManifest,
    'bundleFingerprint'
  >;
  let compiledSizeGuess = 0;
  let finalManifestBytes: Uint8Array | null = null;
  for (let iter = 0; iter < 16; iter++) {
    const candidate: SerializedManifestShape = {
      ...baseManifest,
      compiledSize: compiledSizeGuess,
    };
    const bytes = encoder.encode(stableStringify(candidate));
    const total =
      OFFSET_MANIFEST_BODY + bytes.length + packedPayloadBytes.length;
    if (total === compiledSizeGuess) {
      finalManifestBytes = bytes;
      break;
    }
    compiledSizeGuess = total;
  }
  if (finalManifestBytes === null) {
    throw new Error(
      'mesh-policy-compile: compiledSize fixed-point did not converge',
    );
  }
  const total =
    OFFSET_MANIFEST_BODY + finalManifestBytes.length + packedPayloadBytes.length;
  const bwBytes = new Uint8Array(total);
  bwBytes.set(MAGIC_BYTES, OFFSET_MAGIC);
  bwBytes[OFFSET_VERSION] = CONTAINER_VERSION;
  bwBytes[OFFSET_KIND] = KIND_MESH_POLICY;
  writeU32LE(bwBytes, OFFSET_MANIFEST_LEN, finalManifestBytes.length);
  bwBytes.set(finalManifestBytes, OFFSET_MANIFEST_BODY);
  bwBytes.set(
    packedPayloadBytes,
    OFFSET_MANIFEST_BODY + finalManifestBytes.length,
  );

  // Content-address the fully-assembled bundle. Because `bwBytes` does
  // not itself embed the fingerprint (the serialized manifest JSON
  // carries only `baseManifest` + `compiledSize`), this digest is
  // well-defined — no self-reference cycle. By `compile_deterministic`
  // in MeshPolicyCompileCorrectness.lean, equal input policies produce
  // equal `bwBytes`, hence equal fingerprints.
  const bundleFingerprint = fingerprintMeshPolicyBundle(bwBytes);

  const finalManifest: MeshPolicyCompiledManifest = {
    ...baseManifest,
    compiledSize: compiledSizeGuess,
    bundleFingerprint,
  };

  return { bwBytes, manifest: finalManifest };
}

// ─── Load ────────────────────────────────────────────────────────────

function parseManifest(
  raw: unknown,
  bundleFingerprint: string,
): MeshPolicyCompiledManifest {
  if (raw === null || typeof raw !== 'object') {
    throw new Error('mesh-policy-compile: manifest is not an object');
  }
  const m = raw as Record<string, unknown>;
  const format = m.format;
  const portCount = m.portCount;
  const overlayCount = m.overlayCount;
  const authFlag = m.authFlag;
  const modeCode = m.modeCode;
  const honeypitPresent = m.honeypitPresent;
  const compiledAt = m.compiledAt;
  const nominalSize = m.nominalSize;
  const compiledSize = m.compiledSize;
  const packedLength = m.packedLength;
  const parityBytes = m.parityBytes;
  const tailBytes = m.tailBytes;
  if (typeof format !== 'string') {
    throw new Error('mesh-policy-compile: manifest.format not a string');
  }
  if (typeof portCount !== 'number' || !Number.isFinite(portCount)) {
    throw new Error('mesh-policy-compile: manifest.portCount invalid');
  }
  if (typeof overlayCount !== 'number' || !Number.isFinite(overlayCount)) {
    throw new Error('mesh-policy-compile: manifest.overlayCount invalid');
  }
  if (typeof authFlag !== 'boolean') {
    throw new Error('mesh-policy-compile: manifest.authFlag invalid');
  }
  if (typeof modeCode !== 'number' || !Number.isFinite(modeCode)) {
    throw new Error('mesh-policy-compile: manifest.modeCode invalid');
  }
  if (typeof honeypitPresent !== 'boolean') {
    throw new Error('mesh-policy-compile: manifest.honeypitPresent invalid');
  }
  if (typeof compiledAt !== 'string') {
    throw new Error('mesh-policy-compile: manifest.compiledAt invalid');
  }
  if (typeof nominalSize !== 'number' || !Number.isFinite(nominalSize)) {
    throw new Error('mesh-policy-compile: manifest.nominalSize invalid');
  }
  if (typeof compiledSize !== 'number' || !Number.isFinite(compiledSize)) {
    throw new Error('mesh-policy-compile: manifest.compiledSize invalid');
  }
  if (typeof packedLength !== 'number' || !Number.isFinite(packedLength)) {
    throw new Error('mesh-policy-compile: manifest.packedLength invalid');
  }
  if (typeof parityBytes !== 'number' || !Number.isFinite(parityBytes)) {
    throw new Error('mesh-policy-compile: manifest.parityBytes invalid');
  }
  if (typeof tailBytes !== 'number' || !Number.isFinite(tailBytes)) {
    throw new Error('mesh-policy-compile: manifest.tailBytes invalid');
  }
  // `format` sentinel is preserved as-read so unknown-version bundles
  // can be inspected and surfaced by the consumer runtime's version
  // gate — parsing does not reject unknown formats here.
  return {
    format: format as typeof MANIFEST_FORMAT,
    portCount,
    overlayCount,
    authFlag,
    modeCode,
    honeypitPresent,
    compiledAt,
    nominalSize,
    compiledSize,
    packedLength,
    parityBytes,
    tailBytes,
    bundleFingerprint,
  };
}

function extractManifestBytes(bwBytes: Uint8Array): {
  readonly manifestBytes: Uint8Array;
  readonly manifestLen: number;
} {
  verifyContainerHeader(bwBytes);
  const manifestLen = readU32LE(bwBytes, OFFSET_MANIFEST_LEN);
  const manifestEnd = OFFSET_MANIFEST_BODY + manifestLen;
  if (manifestEnd > bwBytes.length) {
    throw new RangeError(
      `mesh-policy-compile: manifest length ${manifestLen} overruns bundle (total ${bwBytes.length})`,
    );
  }
  const manifestBytes = bwBytes.subarray(OFFSET_MANIFEST_BODY, manifestEnd);
  return { manifestBytes, manifestLen };
}

/**
 * Extract just the manifest from a `.bw` bundle, without decoding the
 * full policy payload. Useful for tooling (`x-ranger mesh audit
 * --compiled`) and version gates.
 *
 * @throws `Error` when the header is malformed or the manifest JSON
 *         fails to parse.
 */
export function inspectCompiledBundle(
  bwBytes: Uint8Array,
): MeshPolicyCompiledManifest {
  const { manifestBytes } = extractManifestBytes(bwBytes);
  const decoder = new TextDecoder('utf-8', { fatal: true });
  let manifestJson: string;
  try {
    manifestJson = decoder.decode(manifestBytes);
  } catch (_err) {
    throw new Error('mesh-policy-compile: manifest bytes are not valid UTF-8');
  }
  let raw: unknown;
  try {
    raw = JSON.parse(manifestJson);
  } catch (_err) {
    throw new Error('mesh-policy-compile: manifest JSON parse error');
  }
  const bundleFingerprint = fingerprintMeshPolicyBundle(bwBytes);
  return parseManifest(raw, bundleFingerprint);
}

/**
 * Inverse of {@link compileMeshPolicy}. Returns the reconstructed
 * policy AND the embedded manifest. Mirrors the Lean `loadMeshPolicy`
 * function; paired with compile it satisfies
 * `loadMeshPolicyFromBundle(compileMeshPolicy(p).bwBytes).policy === p`
 * structurally.
 */
export function loadMeshPolicyFromBundle(
  bwBytes: Uint8Array,
): { readonly policy: MeshPolicyConfig; readonly manifest: MeshPolicyCompiledManifest } {
  const { manifestBytes, manifestLen } = extractManifestBytes(bwBytes);
  const decoder = new TextDecoder('utf-8', { fatal: true });

  let manifestJson: string;
  try {
    manifestJson = decoder.decode(manifestBytes);
  } catch (_err) {
    throw new Error('mesh-policy-compile: manifest bytes are not valid UTF-8');
  }
  let rawManifest: unknown;
  try {
    rawManifest = JSON.parse(manifestJson);
  } catch (_err) {
    throw new Error('mesh-policy-compile: manifest JSON parse error');
  }
  const bundleFingerprint = fingerprintMeshPolicyBundle(bwBytes);
  const manifest = parseManifest(rawManifest, bundleFingerprint);

  const payloadStart = OFFSET_MANIFEST_BODY + manifestLen;
  const packedPayloadBytes = bwBytes.subarray(payloadStart);

  // Reverse the opaque-bytes framing. The manifest carries the
  // originalLength + packedLength + parityBytes + tailBytes fields
  // required to reconstruct the original policy JSON byte-for-byte.
  const opaqueManifest: OpaquePackManifest = {
    originalLength: manifest.nominalSize,
    packedLength: manifest.packedLength,
    parityBytes: manifest.parityBytes,
    tailBytes: manifest.tailBytes,
  };

  let payloadBytes: Uint8Array;
  try {
    payloadBytes = unpackOpaqueBytes(packedPayloadBytes, opaqueManifest);
  } catch (err) {
    throw new Error(
      `mesh-policy-compile: payload unpack failed: ${(err as Error).message}`,
    );
  }

  let payloadJson: string;
  try {
    payloadJson = decoder.decode(payloadBytes);
  } catch (_err) {
    throw new Error('mesh-policy-compile: payload bytes are not valid UTF-8');
  }
  let rawPayload: unknown;
  try {
    rawPayload = JSON.parse(payloadJson);
  } catch (_err) {
    throw new Error('mesh-policy-compile: payload JSON parse error');
  }
  if (rawPayload === null || typeof rawPayload !== 'object') {
    throw new Error('mesh-policy-compile: payload is not an object');
  }
  const p = rawPayload as SerializedPolicyPayload;
  if (!Array.isArray(p.ports)) {
    throw new Error('mesh-policy-compile: payload.ports missing or not an array');
  }

  const policy = payloadToPolicy(p);
  return { policy, manifest };
}

// ─── Fingerprint (sync pure-JS SHA-256) ──────────────────────────────

/** SHA-256 round constants (first 32 bits of the fractional parts of the
 *  cube roots of the first 64 primes). Frozen — never mutated. */
const SHA256_K: readonly number[] = [
  0x428a2f98, 0x71374491, 0xb5c0fbcf, 0xe9b5dba5, 0x3956c25b, 0x59f111f1,
  0x923f82a4, 0xab1c5ed5, 0xd807aa98, 0x12835b01, 0x243185be, 0x550c7dc3,
  0x72be5d74, 0x80deb1fe, 0x9bdc06a7, 0xc19bf174, 0xe49b69c1, 0xefbe4786,
  0x0fc19dc6, 0x240ca1cc, 0x2de92c6f, 0x4a7484aa, 0x5cb0a9dc, 0x76f988da,
  0x983e5152, 0xa831c66d, 0xb00327c8, 0xbf597fc7, 0xc6e00bf3, 0xd5a79147,
  0x06ca6351, 0x14292967, 0x27b70a85, 0x2e1b2138, 0x4d2c6dfc, 0x53380d13,
  0x650a7354, 0x766a0abb, 0x81c2c92e, 0x92722c85, 0xa2bfe8a1, 0xa81a664b,
  0xc24b8b70, 0xc76c51a3, 0xd192e819, 0xd6990624, 0xf40e3585, 0x106aa070,
  0x19a4c116, 0x1e376c08, 0x2748774c, 0x34b0bcb5, 0x391c0cb3, 0x4ed8aa4a,
  0x5b9cca4f, 0x682e6ff3, 0x748f82ee, 0x78a5636f, 0x84c87814, 0x8cc70208,
  0x90befffa, 0xa4506ceb, 0xbef9a3f7, 0xc67178f2,
];

/** SHA-256 initial hash state (first 32 bits of the fractional parts of
 *  the square roots of the first 8 primes). */
const SHA256_H0: readonly number[] = [
  0x6a09e667, 0xbb67ae85, 0x3c6ef372, 0xa54ff53a, 0x510e527f, 0x9b05688c,
  0x1f83d9ab, 0x5be0cd19,
];

function rotr32(x: number, n: number): number {
  return ((x >>> n) | (x << (32 - n))) >>> 0;
}

function add32(...xs: number[]): number {
  let sum = 0;
  for (const x of xs) sum = (sum + x) >>> 0;
  return sum >>> 0;
}

/** Pure-JS SHA-256 over a byte buffer. Returns the 32-byte digest.
 *  Uses only `Uint8Array` + `DataView` + bitwise ops — no Web Crypto,
 *  no `node:crypto`, no `Buffer`. Keeps the module Node-free for the
 *  runtime surface scan in
 *  `mesh-policy-compile.test.ts :: 'production source files have no
 *  node: imports ...'`. */
function sha256(bytes: Uint8Array): Uint8Array {
  const bitLength = bytes.length * 8;
  // Pad so `(len + 1 + pad + 8) mod 64 === 0`.
  const padBytes = (64 - ((bytes.length + 9) % 64)) % 64;
  const padded = new Uint8Array(bytes.length + 1 + padBytes + 8);
  padded.set(bytes);
  padded[bytes.length] = 0x80;
  const padView = new DataView(padded.buffer, padded.byteOffset, padded.byteLength);
  const high = Math.floor(bitLength / 0x100000000);
  const low = bitLength >>> 0;
  padView.setUint32(padded.length - 8, high, false);
  padView.setUint32(padded.length - 4, low, false);

  const state = [...SHA256_H0];
  const w = new Uint32Array(64);

  for (let off = 0; off < padded.length; off += 64) {
    for (let i = 0; i < 16; i++) {
      w[i] = padView.getUint32(off + i * 4, false);
    }
    for (let i = 16; i < 64; i++) {
      const s0 = rotr32(w[i - 15]!, 7) ^ rotr32(w[i - 15]!, 18) ^ (w[i - 15]! >>> 3);
      const s1 = rotr32(w[i - 2]!, 17) ^ rotr32(w[i - 2]!, 19) ^ (w[i - 2]! >>> 10);
      w[i] = add32(w[i - 16]!, s0, w[i - 7]!, s1);
    }
    let [a, b, c, d, e, f, g, h] = state;
    for (let i = 0; i < 64; i++) {
      const S1 = rotr32(e!, 6) ^ rotr32(e!, 11) ^ rotr32(e!, 25);
      const ch = (e! & f!) ^ (~e! & g!);
      const t1 = add32(h!, S1, ch, SHA256_K[i]!, w[i]!);
      const S0 = rotr32(a!, 2) ^ rotr32(a!, 13) ^ rotr32(a!, 22);
      const maj = (a! & b!) ^ (a! & c!) ^ (b! & c!);
      const t2 = add32(S0, maj);
      h = g;
      g = f;
      f = e;
      e = add32(d!, t1);
      d = c;
      c = b;
      b = a;
      a = add32(t1, t2);
    }
    state[0] = add32(state[0]!, a!);
    state[1] = add32(state[1]!, b!);
    state[2] = add32(state[2]!, c!);
    state[3] = add32(state[3]!, d!);
    state[4] = add32(state[4]!, e!);
    state[5] = add32(state[5]!, f!);
    state[6] = add32(state[6]!, g!);
    state[7] = add32(state[7]!, h!);
  }

  const out = new Uint8Array(32);
  const outView = new DataView(out.buffer);
  for (let i = 0; i < 8; i++) {
    outView.setUint32(i * 4, state[i]!, false);
  }
  return out;
}

/**
 * Content-addressed fingerprint of a compiled mesh-policy bundle.
 *
 * Returns a 64-character lowercase-hex SHA-256 digest of `bwBytes`.
 * Deterministic: equal input policies compile to equal `bwBytes` (by
 * `MeshPolicyCompileCorrectness.compile_deterministic`), hence equal
 * fingerprints. Use this for idempotent deploy tooling — compare the
 * locally-computed fingerprint against the currently-deployed one and
 * skip the round-trip when they match.
 */
export function fingerprintMeshPolicyBundle(bwBytes: Uint8Array): string {
  const digest = sha256(bwBytes);
  let hex = '';
  for (let i = 0; i < digest.length; i++) {
    const byte = digest[i]!;
    hex += (byte < 16 ? '0' : '') + byte.toString(16);
  }
  return hex;
}

// ─── Re-exports for consumer convenience ─────────────────────────────

export {
  MAGIC_BYTES as MESH_POLICY_BW_MAGIC,
  CONTAINER_VERSION as MESH_POLICY_BW_VERSION,
  KIND_MESH_POLICY as MESH_POLICY_BW_KIND,
  MANIFEST_FORMAT as MESH_POLICY_MANIFEST_FORMAT,
};
