// @a0n/gnosis/src/module-system.ts
/**
 * module-system.ts -- Reproducible build units with real distribution
 *
 * Lockfile-backed dependency resolution, module contracts with effect
 * signatures, and version constraint checking.
 *
 * Roadmap item #7: ships each new feature with reproducible package behavior.
 */

import type { EffectContract, EffectKind } from './effects';

// ============================================================================
// Version constraints
// ============================================================================

export interface SemVer {
  major: number;
  minor: number;
  patch: number;
  prerelease?: string;
}

/**
 * Parses the Sem Ver.
 */
export function parseSemVer(version: string): SemVer | null {
  const match = version.match(/^(\d+)\.(\d+)\.(\d+)(?:-(.+))?$/);
  if (!match) return null;
  return {
    major: parseInt(match[1], 10),
    minor: parseInt(match[2], 10),
    patch: parseInt(match[3], 10),
    ...(match[4] ? { prerelease: match[4] } : {}),
  };
}

/**
 * Formats the Sem Ver.
 */
export function formatSemVer(v: SemVer): string {
  return `${v.major}.${v.minor}.${v.patch}${
    v.prerelease ? `-${v.prerelease}` : ''
  }`;
}

/**
 * Handles the gnosis compare Sem Ver workflow.
 */
export function compareSemVer(a: SemVer, b: SemVer): number {
  if (a.major !== b.major) return a.major - b.major;
  if (a.minor !== b.minor) return a.minor - b.minor;
  if (a.patch !== b.patch) return a.patch - b.patch;
  if (a.prerelease && !b.prerelease) return -1;
  if (!a.prerelease && b.prerelease) return 1;
  return 0;
}

export type VersionConstraint =
  | { type: 'exact'; version: SemVer }
  | { type: 'caret'; version: SemVer } // ^1.2.3 (compatible)
  | { type: 'tilde'; version: SemVer } // ~1.2.3 (patch-level)
  | { type: 'range'; min: SemVer; max: SemVer }
  | { type: 'any' };

/**
 * Parses the Constraint.
 */
export function parseConstraint(raw: string): VersionConstraint | null {
  if (raw === '*') return { type: 'any' };

  if (raw.startsWith('^')) {
    const v = parseSemVer(raw.slice(1));
    return v ? { type: 'caret', version: v } : null;
  }

  if (raw.startsWith('~')) {
    const v = parseSemVer(raw.slice(1));
    return v ? { type: 'tilde', version: v } : null;
  }

  const v = parseSemVer(raw);
  return v ? { type: 'exact', version: v } : null;
}

/**
 * Handles the gnosis satisfies Constraint workflow.
 */
export function satisfiesConstraint(
  version: SemVer,
  constraint: VersionConstraint
): boolean {
  switch (constraint.type) {
    case 'any':
      return true;

    case 'exact':
      return compareSemVer(version, constraint.version) === 0;

    case 'caret': {
      // ^1.2.3: >=1.2.3, <2.0.0 (for major > 0)
      // ^0.2.3: >=0.2.3, <0.3.0
      const cv = constraint.version;
      if (compareSemVer(version, cv) < 0) return false;
      if (cv.major > 0) return version.major === cv.major;
      if (cv.minor > 0)
        return version.major === cv.major && version.minor === cv.minor;
      return compareSemVer(version, cv) === 0;
    }

    case 'tilde': {
      // ~1.2.3: >=1.2.3, <1.3.0
      const cv = constraint.version;
      if (compareSemVer(version, cv) < 0) return false;
      return version.major === cv.major && version.minor === cv.minor;
    }

    case 'range':
      return (
        compareSemVer(version, constraint.min) >= 0 &&
        compareSemVer(version, constraint.max) <= 0
      );
  }
}

// ============================================================================
// Module manifest -- what a module declares about itself
// ============================================================================

export interface ModuleDependency {
  name: string;
  constraint: VersionConstraint;
  /** Whether this is a dev-only dependency */
  dev: boolean;
}

export interface ModuleManifest {
  /** Module name (scoped, e.g., '@gnosis/void-walker') */
  name: string;
  /** Module version */
  version: SemVer;
  /** Dependencies */
  dependencies: ModuleDependency[];
  /** Exported topology names */
  exports: string[];
  /** Effect contract */
  effects: EffectContract | null;
  /** Minimum gnosis version required */
  gnosisVersion?: VersionConstraint;
}

/**
 * Creates the Manifest.
 */
export function createManifest(name: string, version: string): ModuleManifest {
  const v = parseSemVer(version);
  if (!v) throw new Error(`Invalid version: ${version}`);
  return {
    name,
    version: v,
    dependencies: [],
    exports: [],
    effects: null,
  };
}

// ============================================================================
// Lockfile -- reproducible dependency resolution
// ============================================================================

export interface LockfileEntry {
  /** Module name */
  name: string;
  /** Resolved version */
  version: SemVer;
  /** Integrity hash (content-addressable) */
  integrity: string;
  /** Resolved dependency versions */
  dependencies: Record<string, string>;
}

export interface Lockfile {
  /** Lockfile format version */
  lockfileVersion: number;
  /** Resolved entries keyed by "name@version" */
  entries: Map<string, LockfileEntry>;
}

/**
 * Creates the Lockfile.
 */
export function createLockfile(): Lockfile {
  return { lockfileVersion: 1, entries: new Map() };
}

/**
 * Handles the gnosis add Lockfile Entry workflow.
 */
export function addLockfileEntry(
  lockfile: Lockfile,
  entry: LockfileEntry
): void {
  const key = `${entry.name}@${formatSemVer(entry.version)}`;
  lockfile.entries.set(key, entry);
}

/**
 * Handles the gnosis lockfile Has workflow.
 */
export function lockfileHas(
  lockfile: Lockfile,
  name: string,
  version: SemVer
): boolean {
  return lockfile.entries.has(`${name}@${formatSemVer(version)}`);
}

/**
 * Handles the gnosis lockfile Get workflow.
 */
export function lockfileGet(
  lockfile: Lockfile,
  name: string,
  version: SemVer
): LockfileEntry | undefined {
  return lockfile.entries.get(`${name}@${formatSemVer(version)}`);
}

// ============================================================================
// Dependency resolver -- topological sort with constraint satisfaction
// ============================================================================

export interface ResolutionResult {
  /** Whether all dependencies resolved */
  ok: boolean;
  /** Resolved modules in install order */
  resolved: LockfileEntry[];
  /** Unresolved dependencies */
  unresolved: { name: string; constraint: string; reason: string }[];
  /** Conflicts (multiple incompatible versions) */
  conflicts: { name: string; versions: string[]; constraints: string[] }[];
}

/**
 * Resolve dependencies from a set of manifests and available versions.
 * Uses a simple greedy algorithm: pick the highest version that satisfies
 * all constraints.
 */
export function resolveDependencies(
  root: ModuleManifest,
  available: Map<string, SemVer[]>
): ResolutionResult {
  const resolved: LockfileEntry[] = [];
  const unresolved: ResolutionResult['unresolved'] = [];
  const conflicts: ResolutionResult['conflicts'] = [];
  const picked = new Map<string, SemVer>();

  // Collect all constraints
  const constraints = new Map<string, VersionConstraint[]>();
  for (const dep of root.dependencies) {
    const existing = constraints.get(dep.name) ?? [];
    existing.push(dep.constraint);
    constraints.set(dep.name, existing);
  }

  // Resolve each dependency
  for (const [name, depConstraints] of constraints) {
    const versions = available.get(name);
    if (!versions || versions.length === 0) {
      unresolved.push({
        name,
        constraint: depConstraints.map((c) => constraintToString(c)).join(', '),
        reason: 'No versions available',
      });
      continue;
    }

    // Sort descending (highest first)
    const sorted = [...versions].sort((a, b) => -compareSemVer(a, b));

    // Find highest version satisfying all constraints
    const match = sorted.find((v) =>
      depConstraints.every((c) => satisfiesConstraint(v, c))
    );

    if (!match) {
      unresolved.push({
        name,
        constraint: depConstraints.map((c) => constraintToString(c)).join(', '),
        reason: 'No version satisfies all constraints',
      });
      continue;
    }

    picked.set(name, match);
    resolved.push({
      name,
      version: match,
      integrity: computeIntegrity(name, match),
      dependencies: {},
    });
  }

  return {
    ok: unresolved.length === 0 && conflicts.length === 0,
    resolved,
    unresolved,
    conflicts,
  };
}

function constraintToString(c: VersionConstraint): string {
  switch (c.type) {
    case 'any':
      return '*';
    case 'exact':
      return formatSemVer(c.version);
    case 'caret':
      return `^${formatSemVer(c.version)}`;
    case 'tilde':
      return `~${formatSemVer(c.version)}`;
    case 'range':
      return `${formatSemVer(c.min)}-${formatSemVer(c.max)}`;
  }
}

function computeIntegrity(name: string, version: SemVer): string {
  // Deterministic hash placeholder -- real implementation would use content hash
  const input = `${name}@${formatSemVer(version)}`;
  let hash = 0;
  for (let i = 0; i < input.length; i++) {
    hash = ((hash << 5) - hash + input.charCodeAt(i)) | 0;
  }
  return `sha256-${Math.abs(hash).toString(16).padStart(8, '0')}`;
}

// ============================================================================
// Module compatibility -- effect-aware composition
// ============================================================================

export interface CompatibilityReport {
  compatible: boolean;
  missingEffects: EffectKind[];
  versionConflicts: { name: string; required: string; available: string }[];
}

/**
 * Handles the gnosis check Module Compatibility workflow.
 */
export function checkModuleCompatibility(
  consumer: ModuleManifest,
  provider: ModuleManifest
): CompatibilityReport {
  const missingEffects: EffectKind[] = [];
  const versionConflicts: CompatibilityReport['versionConflicts'] = [];

  // Check effect compatibility
  if (consumer.effects && provider.effects) {
    for (const required of consumer.effects.requires) {
      if (!provider.effects.provides.includes(required)) {
        missingEffects.push(required);
      }
    }
  }

  // Check version constraints
  for (const dep of consumer.dependencies) {
    if (dep.name === provider.name) {
      if (!satisfiesConstraint(provider.version, dep.constraint)) {
        versionConflicts.push({
          name: provider.name,
          required: constraintToString(dep.constraint),
          available: formatSemVer(provider.version),
        });
      }
    }
  }

  return {
    compatible: missingEffects.length === 0 && versionConflicts.length === 0,
    missingEffects,
    versionConflicts,
  };
}
