// @a0n/gnosis/src/neo4j-bridge.ts
import { BettyCompiler } from './betty/compiler';

/**
 * Options for Gnosis to Neo4j export.
 */
export interface GnosisNeo4jOptions {
  /** Label for nodes (default: "GnosisNode") */
  nodeLabel?: string;
  /** Prefix for IDs to avoid collisions */
  idPrefix?: string;
}

/**
 * Bridge between Gnosis and Neo4j.
 */
export class GnosisNeo4jBridge {
  private compiler: BettyCompiler;

  constructor() {
    this.compiler = new BettyCompiler();
  }

  /**
   * Parses GGL code and returns Cypher statements to represent it in Neo4j.
   */
  public gglToCypher(ggl: string, options: GnosisNeo4jOptions = {}): string {
    const { ast, diagnostics } = this.compiler.parse(ggl);
    if (!ast) return '// No AST generated';

    const nodeLabel = options.nodeLabel || 'GnosisNode';
    const idPrefix = options.idPrefix || '';
    const lines: string[] = ['// Gnosis GGL to Neo4j Export'];

    if (diagnostics.some((d) => d.severity === 'error')) {
      lines.push(`// WARNING: Compilation had errors`);
    }

    // Nodes
    for (const [id, node] of ast.nodes) {
      const fullId = `${idPrefix}${id}`;
      const labels = [nodeLabel, ...node.labels].join(':');
      const props = {
        id: fullId,
        originalId: id,
        ...node.properties,
      };
      const propsStr = this.propsToCypher(props);
      lines.push(
        `MERGE (n:${labels} {id: '${fullId}'}) ON CREATE SET n = ${propsStr} ON MATCH SET n = ${propsStr};`
      );
    }

    // Edges (Gnosis edges can be multi-source/multi-target)
    ast.edges.forEach((edge, index) => {
      const edgeType = edge.type;
      const propsStr = this.propsToCypher({
        edgeIndex: index,
        ...edge.properties,
      });

      edge.sourceIds.forEach((sourceId) => {
        const fullSourceId = `${idPrefix}${sourceId}`;
        edge.targetIds.forEach((targetId) => {
          const fullTargetId = `${idPrefix}${targetId}`;
          lines.push(
            `MATCH (s:${nodeLabel} {id: '${fullSourceId}'}), (t:${nodeLabel} {id: '${fullTargetId}'}) MERGE (s)-[r:${edgeType} ${propsStr}]->(t);`
          );
        });
      });
    });

    return lines.join('\n');
  }

  private propsToCypher(obj: Record<string, any>): string {
    const parts = Object.entries(obj).map(([k, v]) => {
      const val = typeof v === 'string' ? `'${v.replace(/'/g, "\\'")}'` : v;
      return `${k}: ${val}`;
    });
    return `{ ${parts.join(', ')} }`;
  }
}
