// @a0n/gnosis/src/neural-compat.ts
import {
  checkGgProgram,
  getGgRootNodeIds,
  getGgTerminalNodeIds,
  parseGgProgram,
  type GgProgram,
} from '@a0n/aeon-logic';
import { Pipeline } from '@a0n/aeon-pipelines';
import { dirname } from 'node:path';
import { lowerUfcsSource } from './ufcs';
import { loadGnosisModuleFromFile } from './mod/loader';
import {
  createCommandFabricBackendsFromEnv,
  runHeteroMoAFabric,
  type HeteroMoAFabricBackendDescriptor,
  type HeteroMoAFabricCommunityMemory,
  type HeteroMoAFabricPlan,
  type HeteroMoAFabricRunOptions,
  type HeteroMoAFabricRunResult,
} from './runtime/hetero-fabric';
import type { StabilityHeteroMoAFabricWitness } from './betty/stability';
import type { QDoc } from './crdt/qdoc';

export interface Neuron {
  id: string;
  type: 'input' | 'hidden' | 'output' | 'cloud';
  bias: number;
  activation: string;
}

export interface Synapse {
  id: string;
  from_id: string;
  to_id: string;
  weight: number;
}

export interface AdapterTrainingConfig {
  rank: number;
  basePrecision: 'int8' | 'fp16' | 'fp32';
  adapterPrecision: 'fp16' | 'fp32';
  microBatchSize: number;
  idleFlushMs: number;
}

export interface NeuralGraphData {
  nodeCount: number;
  nodes: Array<{
    id: string;
    index: number;
    type: Neuron['type'];
  }>;
  edges: Array<{
    id: string;
    source: number;
    target: number;
    weight: number;
  }>;
}

const DEFAULT_TRAINING_IGNORE_TARGET = -999;
const GG_FILE_EXTENSIONS = ['.gg', '.ggx', '.mgg'];

export interface LoadTopologyOptions {
  preserveWeights?: boolean;
}

interface NeuralHeteroFabricCommonOptions
  extends Omit<
    HeteroMoAFabricRunOptions,
    'plan' | 'witness' | 'inputDocument' | 'communityMemory'
  > {
  backends?: HeteroMoAFabricBackendDescriptor[];
  includeEnvCommandBackends?: boolean;
  inputDocument?: QDoc | null;
  communityMemory?: HeteroMoAFabricCommunityMemory | null;
}

export interface NeuralHeteroFabricPlanOptions
  extends NeuralHeteroFabricCommonOptions {
  plan?: Partial<HeteroMoAFabricPlan>;
  witness?: never;
}

export interface NeuralHeteroFabricWitnessOptions
  extends NeuralHeteroFabricCommonOptions {
  witness: StabilityHeteroMoAFabricWitness;
  plan?: never;
}

export type NeuralHeteroFabricOptions =
  | NeuralHeteroFabricPlanOptions
  | NeuralHeteroFabricWitnessOptions;

interface WebGpuNavigatorLike {
  gpu?: {
    requestAdapter?: () =>
      | Promise<WebGpuAdapterLike | null>
      | WebGpuAdapterLike
      | null;
  };
}

interface WebGpuAdapterLike {
  requestDevice?: () => Promise<WebGpuDeviceLike> | WebGpuDeviceLike;
}

interface WebGpuBufferLike {
  mapAsync?: (mode: number) => Promise<void>;
  getMappedRange?: () => ArrayBuffer;
  unmap?: () => void;
  destroy?: () => void;
}

interface WebGpuQueueLike {
  writeBuffer?: (
    buffer: WebGpuBufferLike,
    bufferOffset: number,
    data: ArrayBuffer | ArrayBufferView
  ) => void;
  submit?: (commands: readonly unknown[]) => void;
}

interface WebGpuComputePassEncoderLike {
  setPipeline?: (pipeline: unknown) => void;
  setBindGroup?: (index: number, bindGroup: unknown) => void;
  dispatchWorkgroups?: (x: number) => void;
  end?: () => void;
}

interface WebGpuCommandEncoderLike {
  beginComputePass?: () => WebGpuComputePassEncoderLike;
  copyBufferToBuffer?: (
    source: WebGpuBufferLike,
    sourceOffset: number,
    destination: WebGpuBufferLike,
    destinationOffset: number,
    size: number
  ) => void;
  finish?: () => unknown;
}

interface WebGpuDeviceLike {
  queue: WebGpuQueueLike;
  createBuffer?: (descriptor: {
    size: number;
    usage: number;
    mappedAtCreation?: boolean;
  }) => WebGpuBufferLike;
  createShaderModule?: (descriptor: { code: string }) => unknown;
  createBindGroupLayout?: (descriptor: unknown) => unknown;
  createPipelineLayout?: (descriptor: unknown) => unknown;
  createComputePipeline?: (descriptor: unknown) => unknown;
  createBindGroup?: (descriptor: unknown) => unknown;
  createCommandEncoder?: () => WebGpuCommandEncoderLike;
}

export const TOPIC_DOMAIN_TRANSFORMER_TOPOLOGY_FILE =
  'topic_domain_transformer.gg';

const TOPIC_DOMAIN_TRANSFORMER_TOPOLOGY_ASSET_RELATIVE_PATH = `../${TOPIC_DOMAIN_TRANSFORMER_TOPOLOGY_FILE}`;

// This inline source is the browser/runtime fallback for the canonical module
// defined at open-source/gnosis/topic_domain_transformer.gg.
export const TOPIC_DOMAIN_TRANSFORMER_TOPOLOGY = `
(topic_tokens: Tensor { role: 'input', activation: 'identity', bias: '0.0' })
(topic_embedding: Projection { role: 'embedding', activation: 'identity', bias: '0.0' })

(qkv_projection: Projection { role: 'qkv_projection', activation: 'identity', bias: '0.0' })

(head_a: AttentionHead { bias: '0.02' })
(head_b: AttentionHead { bias: '0.02' })
(head_c: AttentionHead { bias: '0.02' })
(head_d: AttentionHead { bias: '0.02' })

(attention_mix: Tensor { role: 'multi_head_mix', activation: 'tanh', bias: '0.03' })
(topic_state: Tensor { role: 'residual_state', activation: 'tanh', bias: '0.01' })
(topic_logits: Tensor { role: 'logits', activation: 'identity', bias: '0.0' })
(topic_distribution: Tensor { role: 'distribution', activation: 'tanh', bias: '0.0' })

(topic_tokens)-[:PROCESS]->(topic_embedding)
(topic_embedding)-[:PROCESS]->(qkv_projection)
(qkv_projection)-[:FORK]->(head_a | head_b | head_c | head_d)

(head_a | head_b | head_c | head_d)-[:FOLD { strategy: 'concat', weight: '0.25' }]->(attention_mix)
(attention_mix | topic_embedding)-[:SLIVER { mode: 'constructive', weight: '0.5' }]->(topic_state)
(topic_state)-[:PROCESS]->(topic_logits)
(topic_logits)-[:PROCESS]->(topic_distribution)
`.trim();

// Backwards-compatible alias kept for existing imports.
export const TRANSFORMER_HELLO_WORLD_TOPOLOGY =
  TOPIC_DOMAIN_TRANSFORMER_TOPOLOGY;

type TrainingEvent = { type: 'loss' | 'epoch'; value: number };

interface NeuralStore {
  neurons: Map<string, Neuron>;
  synapses: Map<string, Synapse>;
  semanticDescriptions: Map<string, string>;
}

const DEFAULT_NEURAL_STORE: NeuralStore = {
  neurons: new Map(),
  synapses: new Map(),
  semanticDescriptions: new Map(),
};

interface BunRuntimeLike {
  file(path: string | URL): {
    text(): Promise<string>;
  };
}

function getBunRuntime(): BunRuntimeLike | null {
  const globalLike = globalThis as { Bun?: BunRuntimeLike };
  return typeof globalLike.Bun?.file === 'function' ? globalLike.Bun : null;
}

function isGgFilePath(filePath: string): boolean {
  const lower = filePath.toLowerCase();
  return GG_FILE_EXTENSIONS.some((extension) => lower.endsWith(extension));
}

function formatUnknownError(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

function dynamicImport<ModuleShape>(specifier: string): Promise<ModuleShape> {
  return import(/* @vite-ignore */ specifier) as Promise<ModuleShape>;
}

async function readTopologyFromPath(topologyFilePath: string): Promise<string> {
  if (!isGgFilePath(topologyFilePath)) {
    throw new Error(
      `Topology file must end with ${GG_FILE_EXTENSIONS.join(' or ')}`
    );
  }

  if (topologyFilePath.toLowerCase().endsWith('.mgg')) {
    try {
      const module = await loadGnosisModuleFromFile(
        topologyFilePath,
        dirname(topologyFilePath)
      );
      return module.mergedSource;
    } catch (error) {
      throw new Error(
        `Unable to load .mgg topology module "${topologyFilePath}": ${formatUnknownError(
          error
        )}`
      );
    }
  }

  const bunRuntime = getBunRuntime();
  if (bunRuntime) {
    return bunRuntime.file(topologyFilePath).text();
  }

  try {
    const fsPromises = await dynamicImport<{
      readFile(path: string, encoding: string): Promise<string>;
    }>('node:fs/promises');
    return await fsPromises.readFile(topologyFilePath, 'utf-8');
  } catch (error) {
    throw new Error(
      `Unable to load topology file "${topologyFilePath}": ${formatUnknownError(
        error
      )}`
    );
  }
}

async function tryReadCanonicalTopicDomainTopology(): Promise<string | null> {
  const sourceFileUrl = new URL(
    TOPIC_DOMAIN_TRANSFORMER_TOPOLOGY_ASSET_RELATIVE_PATH,

    import.meta.url
  );

  const bunRuntime = getBunRuntime();
  if (bunRuntime) {
    try {
      return await bunRuntime.file(sourceFileUrl).text();
    } catch {
      return null;
    }
  }

  try {
    const fsPromises = await dynamicImport<{
      readFile(path: string, encoding: string): Promise<string>;
    }>('node:fs/promises');
    const nodeUrl = await dynamicImport<{
      fileURLToPath(url: URL): string;
    }>('node:url');

    return await fsPromises.readFile(
      nodeUrl.fileURLToPath(sourceFileUrl),
      'utf-8'
    );
  } catch {
    return null;
  }
}

/**
 * Handles the gnosis get Topic Domain Transformer Topology Source workflow.
 */
export async function getTopicDomainTransformerTopologySource(): Promise<string> {
  const canonicalSource = await tryReadCanonicalTopicDomainTopology();
  return canonicalSource ?? TOPIC_DOMAIN_TRANSFORMER_TOPOLOGY;
}

function parseNumericProperty(
  raw: string | undefined,
  fallback: number
): number {
  if (!raw) {
    return fallback;
  }
  const parsed = Number.parseFloat(raw);
  return Number.isFinite(parsed) ? parsed : fallback;
}

function normalizeActivation(raw: string | undefined): string {
  if (!raw) {
    return 'tanh';
  }
  const value = raw.trim().toLowerCase();
  if (value === 'identity' || value === 'linear') {
    return 'identity';
  }
  if (value === 'relu') {
    return 'relu';
  }
  return 'tanh';
}

function inferEdgeWeight(
  edgeType: string,
  sourceCount: number,
  targetCount: number
): number {
  const type = edgeType.toUpperCase();
  if (type === 'FORK') {
    return 1 / Math.max(1, targetCount);
  }
  if (type === 'FOLD' || type === 'COLLAPSE') {
    return 1 / Math.max(1, sourceCount);
  }
  if (type === 'RACE') {
    return 0.75;
  }
  if (type === 'SLIVER') {
    return 0.5;
  }
  if (type === 'VENT' || type === 'TUNNEL') {
    return 0;
  }
  return 1;
}

function buildGraphFromGg(program: GgProgram): {
  neurons: Neuron[];
  synapses: Synapse[];
} {
  const rootIds = new Set(getGgRootNodeIds(program));
  const terminalIds = new Set(getGgTerminalNodeIds(program));

  const neurons: Neuron[] = program.nodes.map((node: any) => {
    const isRoot = rootIds.has(node.id);
    const isTerminal = terminalIds.has(node.id);
    const type: Neuron['type'] = isRoot
      ? 'input'
      : isTerminal
      ? 'output'
      : 'hidden';

    return {
      id: node.id,
      type,
      bias: parseNumericProperty(node.properties.bias, 0),
      activation: normalizeActivation(node.properties.activation),
    };
  });

  const synapses: Synapse[] = [];
  let edgeIndex = 0;
  for (const edge of program.edges) {
    const configuredWeight = parseNumericProperty(edge.properties.weight, NaN);
    const edgeWeight = Number.isFinite(configuredWeight)
      ? configuredWeight
      : inferEdgeWeight(
          edge.type,
          edge.sourceIds.length,
          edge.targetIds.length
        );

    for (const sourceId of edge.sourceIds) {
      for (const targetId of edge.targetIds) {
        synapses.push({
          id: `syn-${edge.type.toLowerCase()}-${edgeIndex}`,
          from_id: sourceId,
          to_id: targetId,
          weight: edgeWeight,
        });
        edgeIndex++;
      }
    }
  }

  return { neurons, synapses };
}

export class Translator {
  private idToIndex = new Map<string, number>();

  flatten(
    neurons: Neuron[],
    synapses: Synapse[]
  ): {
    size: number;
    weights: Float32Array;
    biases: Float32Array;
    initialValues: Float32Array;
  } {
    const size = neurons.length;
    this.idToIndex.clear();

    for (let index = 0; index < neurons.length; index++) {
      const neuron = neurons[index];
      if (neuron === undefined) {
        throw new Error('flatten: missing neuron');
      }
      this.idToIndex.set(neuron.id, index);
    }

    const biases = new Float32Array(size);
    const initialValues = new Float32Array(size);
    for (let index = 0; index < neurons.length; index++) {
      const neuron = neurons[index];
      if (neuron === undefined) {
        throw new Error('flatten: missing neuron');
      }
      biases[index] = neuron.bias;
      initialValues[index] = 0;
    }

    const weights = new Float32Array(size * size);
    for (const synapse of synapses) {
      const fromIndex = this.idToIndex.get(synapse.from_id);
      const toIndex = this.idToIndex.get(synapse.to_id);
      if (fromIndex === undefined || toIndex === undefined) {
        continue;
      }
      weights[toIndex * size + fromIndex] = synapse.weight;
    }

    return { size, weights, biases, initialValues };
  }
}

export class NeuronRepository {
  private readonly store: NeuralStore;

  constructor(store: NeuralStore = DEFAULT_NEURAL_STORE) {
    this.store = store;
  }

  async create(neuron: Neuron): Promise<void> {
    this.store.neurons.set(neuron.id, { ...neuron });
  }

  async createWithSemantics(
    neuron: Neuron,
    description: string
  ): Promise<void> {
    await this.create(neuron);
    this.store.semanticDescriptions.set(neuron.id, description);
  }

  async getAll(): Promise<Neuron[]> {
    return Array.from(this.store.neurons.values());
  }

  async delete(id: string): Promise<void> {
    this.store.neurons.delete(id);
    this.store.semanticDescriptions.delete(id);
  }
}

export class SynapseRepository {
  private readonly store: NeuralStore;

  constructor(store: NeuralStore = DEFAULT_NEURAL_STORE) {
    this.store = store;
  }

  async create(synapse: Synapse): Promise<void> {
    this.store.synapses.set(synapse.id, { ...synapse });
  }

  async getAll(): Promise<Synapse[]> {
    return Array.from(this.store.synapses.values());
  }

  async delete(id: string): Promise<void> {
    this.store.synapses.delete(id);
  }
}

export class GPUEngine {
  networkSize = 0;
  batchSize = 1;

  private initialized = false;
  private weights = new Float32Array();
  private biases = new Float32Array();
  private activations: string[] = [];
  private targets = new Float32Array();
  private learningRate = 0.001;
  private latestInputs = new Float32Array();
  private latestOutputs = new Float32Array();
  private readonly subscribers = new Set<(event: TrainingEvent) => void>();
  private readonly translator = new Translator();
  private readonly neuronDefinitions = new Map<string, Neuron>();
  private readonly synapseDefinitions = new Map<string, Synapse>();

  async init(): Promise<void> {
    this.initialized = true;
  }

  prepareBuffers(
    size: number,
    weights: Float32Array,
    biases: Float32Array,
    batchSize = 1
  ): void {
    if (!this.initialized) {
      throw new Error('GPUEngine not initialized');
    }

    if (size <= 0) {
      throw new Error('Network size must be > 0');
    }

    if (weights.length !== size * size) {
      throw new Error(
        `Weight matrix size mismatch. Expected ${size * size}, got ${
          weights.length
        }`
      );
    }

    if (biases.length !== size) {
      throw new Error(
        `Bias vector size mismatch. Expected ${size}, got ${biases.length}`
      );
    }

    this.networkSize = size;
    this.batchSize = Math.max(1, batchSize);
    this.weights = new Float32Array(weights);
    this.biases = new Float32Array(biases);

    const expectedTargetSize = this.networkSize * this.batchSize;
    if (this.targets.length !== expectedTargetSize) {
      this.targets = new Float32Array(expectedTargetSize).fill(
        DEFAULT_TRAINING_IGNORE_TARGET
      );
    }

    if (this.activations.length !== size) {
      this.activations = new Array(size).fill('tanh');
    }
  }

  setActivations(activations: string[]): void {
    this.activations = [...activations];
  }

  prepareTrainingBuffers(targets: Float32Array, learningRate: number): void {
    if (!this.initialized || this.networkSize === 0) {
      throw new Error('GPU not ready for training');
    }

    const expectedTargetSize = this.networkSize * this.batchSize;
    if (targets.length !== expectedTargetSize) {
      throw new Error(
        `Target size mismatch. Expected ${expectedTargetSize}, got ${targets.length}`
      );
    }

    this.targets = new Float32Array(targets);
    this.learningRate = learningRate;
  }

  async runTick(inputs: Float32Array): Promise<Float32Array> {
    if (!this.initialized) {
      throw new Error('GPUEngine not initialized');
    }

    this.compileGraphDefinitionsIfNeeded();

    const expectedInputSize = this.networkSize * this.batchSize;
    if (inputs.length !== expectedInputSize) {
      throw new Error(
        `Input size mismatch. Expected ${expectedInputSize}, got ${inputs.length}`
      );
    }

    this.latestInputs = new Float32Array(inputs);

    const workFns: Array<() => Promise<Float32Array>> = [];
    for (let batchIndex = 0; batchIndex < this.batchSize; batchIndex++) {
      const offset = batchIndex * this.networkSize;
      const batchInput = inputs.slice(offset, offset + this.networkSize);
      workFns.push(async () => this.forwardOneBatch(batchInput));
    }

    const flattened = await Pipeline.from(workFns).fold({
      type: 'merge-all',
      merge: (results: Map<number, Float32Array>) => {
        const ordered = Array.from(results.entries())
          .sort(([left], [right]) => left - right)
          .map(([, value]) => value);
        const output = new Float32Array(this.networkSize * this.batchSize);
        for (let index = 0; index < ordered.length; index++) {
          const batchOutput = ordered[index];
          if (batchOutput === undefined) {
            throw new Error('runTick: missing batch output');
          }
          output.set(batchOutput, index * this.networkSize);
        }
        return output;
      },
    });

    this.latestOutputs = new Float32Array(flattened);
    return new Float32Array(flattened);
  }

  async train(
    inputs: Float32Array,
    targets: Float32Array
  ): Promise<Float32Array> {
    const outputs = await this.runTick(inputs);
    this.prepareTrainingBuffers(targets, this.learningRate);
    const meanLoss = this.applySingleLayerGradientStep();
    this.emit({ type: 'loss', value: meanLoss });
    this.emit({ type: 'epoch', value: 1 });
    return outputs;
  }

  async trainTick(): Promise<void> {
    if (this.latestInputs.length === 0 || this.latestOutputs.length === 0) {
      throw new Error('No forward pass available for trainTick');
    }

    const meanLoss = this.applySingleLayerGradientStep();
    this.emit({ type: 'loss', value: meanLoss });
    this.emit({ type: 'epoch', value: 1 });
  }

  subscribe(callback: (event: TrainingEvent) => void): () => void {
    this.subscribers.add(callback);
    return () => {
      this.subscribers.delete(callback);
    };
  }

  async injectInput(data: Float32Array): Promise<void> {
    if (!this.initialized || this.networkSize === 0) {
      return;
    }

    const expectedLength = this.networkSize * this.batchSize;
    const nextInputs = new Float32Array(expectedLength);
    nextInputs.set(data.slice(0, expectedLength));
    this.latestInputs = nextInputs;
  }

  async createNeuron(neuron: {
    id: string;
    type: Neuron['type'];
    bias: number;
    activation: string;
  }): Promise<void> {
    this.neuronDefinitions.set(neuron.id, {
      id: neuron.id,
      type: neuron.type,
      bias: neuron.bias,
      activation: neuron.activation,
    });
  }

  async createSynapse(synapse: {
    from: string;
    to: string;
    weight: number;
  }): Promise<void> {
    const id = `${synapse.from}->${synapse.to}`;
    this.synapseDefinitions.set(id, {
      id,
      from_id: synapse.from,
      to_id: synapse.to,
      weight: synapse.weight,
    });
  }

  async uploadBuffer(_buffer: unknown, _data: Float32Array): Promise<void> {
    // Compatibility no-op for adapters that expect this method.
  }

  async getWeights(): Promise<Float32Array[]> {
    return [new Float32Array(this.weights)];
  }

  async getBiases(): Promise<Float32Array[]> {
    return [new Float32Array(this.biases)];
  }

  private compileGraphDefinitionsIfNeeded(): void {
    if (
      this.weights.length > 0 ||
      this.neuronDefinitions.size === 0 ||
      !this.initialized
    ) {
      return;
    }

    const neurons = Array.from(this.neuronDefinitions.values());
    const synapses = Array.from(this.synapseDefinitions.values());
    const flattened = this.translator.flatten(neurons, synapses);

    this.prepareBuffers(
      flattened.size,
      flattened.weights,
      flattened.biases,
      this.batchSize
    );

    this.activations = neurons.map((neuron) =>
      normalizeActivation(neuron.activation)
    );
  }

  private activate(value: number, index: number): number {
    const activation = this.activations[index] ?? 'tanh';
    if (activation === 'identity') {
      return value;
    }
    if (activation === 'relu') {
      return Math.max(0, value);
    }
    return Math.tanh(value);
  }

  private forwardOneBatch(input: Float32Array): Float32Array {
    const output = new Float32Array(this.networkSize);

    for (
      let targetNeuronIndex = 0;
      targetNeuronIndex < this.networkSize;
      targetNeuronIndex++
    ) {
      let sum = this.biases[targetNeuronIndex] ?? 0;
      const rowOffset = targetNeuronIndex * this.networkSize;

      for (
        let sourceNeuronIndex = 0;
        sourceNeuronIndex < this.networkSize;
        sourceNeuronIndex++
      ) {
        const weight = this.weights[rowOffset + sourceNeuronIndex] ?? 0;
        sum += weight * (input[sourceNeuronIndex] ?? 0);
      }

      output[targetNeuronIndex] = this.activate(sum, targetNeuronIndex);
    }

    return output;
  }

  private applySingleLayerGradientStep(): number {
    if (
      this.networkSize === 0 ||
      this.latestInputs.length === 0 ||
      this.latestOutputs.length === 0
    ) {
      return 0;
    }

    let totalLoss = 0;
    let validTargetCount = 0;
    const normalizer = this.batchSize <= 0 ? 1 : this.batchSize;

    for (let batchIndex = 0; batchIndex < this.batchSize; batchIndex++) {
      const batchOffset = batchIndex * this.networkSize;

      for (
        let targetNeuronIndex = 0;
        targetNeuronIndex < this.networkSize;
        targetNeuronIndex++
      ) {
        const targetIndex = batchOffset + targetNeuronIndex;
        const targetValue = this.targets[targetIndex];
        if (
          targetValue === undefined ||
          targetValue <= DEFAULT_TRAINING_IGNORE_TARGET + 1
        ) {
          continue;
        }

        const outputValue = this.latestOutputs[targetIndex] ?? 0;
        const error = outputValue - targetValue;
        totalLoss += 0.5 * error * error;
        validTargetCount++;

        const rowOffset = targetNeuronIndex * this.networkSize;
        for (
          let sourceNeuronIndex = 0;
          sourceNeuronIndex < this.networkSize;
          sourceNeuronIndex++
        ) {
          const inputValue =
            this.latestInputs[batchOffset + sourceNeuronIndex] ?? 0;
          const gradient = (error * inputValue) / normalizer;
          const weightIndex = rowOffset + sourceNeuronIndex;
          const currentWeight = this.weights[weightIndex];
          if (currentWeight === undefined) {
            throw new Error('train: missing weight');
          }
          this.weights[weightIndex] = currentWeight - this.learningRate * gradient;
        }

        const currentBias = this.biases[targetNeuronIndex];
        if (currentBias === undefined) {
          throw new Error('train: missing bias');
        }
        this.biases[targetNeuronIndex] = currentBias - this.learningRate * (error / normalizer);
      }
    }

    if (validTargetCount === 0) {
      return 0;
    }

    return totalLoss / validTargetCount;
  }

  private emit(event: TrainingEvent): void {
    for (const subscriber of this.subscribers) {
      subscriber(event);
    }
  }
}

export class WebGPUEngine extends GPUEngine {
  adapter: WebGpuAdapterLike | null = null;
  device: WebGpuDeviceLike | null = null;
  isReady = false;

  private gpuWeights = new Float32Array();
  private gpuBiases = new Float32Array();
  private gpuActivations = new Uint32Array();

  override async init(): Promise<void> {
    await super.init();
    const globalNavigator =
      typeof navigator === 'undefined'
        ? undefined
        : (navigator as WebGpuNavigatorLike);

    const adapter = await globalNavigator?.gpu?.requestAdapter?.();
    if (!adapter?.requestDevice) {
      this.isReady = false;
      return;
    }

    this.adapter = adapter;
    this.device = await adapter.requestDevice();
    this.isReady = this.device !== null;
  }

  override prepareBuffers(
    size: number,
    weights: Float32Array,
    biases: Float32Array,
    batchSize = 1
  ): void {
    super.prepareBuffers(size, weights, biases, batchSize);
    this.gpuWeights = new Float32Array(weights);
    this.gpuBiases = new Float32Array(biases);
  }

  override setActivations(activations: string[]): void {
    super.setActivations(activations);
    this.gpuActivations = new Uint32Array(
      activations.map((activation) => {
        const normalized = normalizeActivation(activation);
        if (normalized === 'relu') {
          return 1;
        }
        if (normalized === 'tanh') {
          return 2;
        }
        return 0;
      })
    );
  }

  override async runTick(inputs: Float32Array): Promise<Float32Array> {
    if (!this.isReady || !this.device) {
      return super.runTick(inputs);
    }

    if (this.networkSize === 0 || this.batchSize <= 0) {
      return super.runTick(inputs);
    }

    const totalSize = this.networkSize * this.batchSize;
    if (inputs.length !== totalSize) {
      throw new Error(
        `Input size mismatch. Expected ${totalSize}, got ${inputs.length}`
      );
    }

    const output = await this.tryRunWebGpu(inputs, totalSize);
    if (output) {
      return output;
    }

    return super.runTick(inputs);
  }

  private resolveBufferUsage(name: string, fallback: number): number {
    const globalUsage = (globalThis as Record<string, unknown>).GPUBufferUsage;
    if (
      typeof globalUsage === 'object' &&
      globalUsage !== null &&
      name in globalUsage &&
      typeof (globalUsage as Record<string, unknown>)[name] === 'number'
    ) {
      const usage = (globalUsage as Record<string, number>)[name];
      if (usage !== undefined) {
        return usage;
      }
    }
    return fallback;
  }

  private resolveMapModeRead(): number {
    const globalMode = (globalThis as Record<string, unknown>).GPUMapMode;
    if (
      typeof globalMode === 'object' &&
      globalMode !== null &&
      'READ' in globalMode &&
      typeof (globalMode as Record<string, unknown>).READ === 'number'
    ) {
      const read = (globalMode as Record<string, number>).READ;
      if (read !== undefined) {
        return read;
      }
    }
    return 1;
  }

  private async tryRunWebGpu(
    inputs: Float32Array,
    totalSize: number
  ): Promise<Float32Array | null> {
    const device = this.device;
    if (
      !device?.createBuffer ||
      !device.createShaderModule ||
      !device.createBindGroupLayout ||
      !device.createPipelineLayout ||
      !device.createComputePipeline ||
      !device.createBindGroup ||
      !device.createCommandEncoder ||
      !device.queue.writeBuffer ||
      !device.queue.submit
    ) {
      return null;
    }

    try {
      const shader = `
struct Meta {
  networkSize : u32,
  totalSize : u32,
  _pad0 : u32,
  _pad1 : u32,
}

@group(0) @binding(0) var<storage, read> weights : array<f32>;
@group(0) @binding(1) var<storage, read> biases : array<f32>;
@group(0) @binding(2) var<storage, read> inputs : array<f32>;
@group(0) @binding(3) var<storage, read> activations : array<u32>;
@group(0) @binding(4) var<storage, read_write> outputs : array<f32>;
@group(0) @binding(5) var<uniform> meta : Meta;

fn tanhApprox(x: f32) -> f32 {
  let doubled = exp(2.0 * x);
  return (doubled - 1.0) / (doubled + 1.0);
}

@compute @workgroup_size(64)
fn main(@builtin(global_invocation_id) gid : vec3<u32>) {
  let index = gid.x;
  if (index >= meta.totalSize) {
    return;
  }

  let batch = index / meta.networkSize;
  let target = index % meta.networkSize;
  let rowOffset = target * meta.networkSize;
  let batchOffset = batch * meta.networkSize;
  var sum = biases[target];

  for (var source: u32 = 0u; source < meta.networkSize; source = source + 1u) {
    sum = sum + weights[rowOffset + source] * inputs[batchOffset + source];
  }

  let activation = activations[target];
  if (activation == 1u) {
    outputs[index] = max(sum, 0.0);
    return;
  }
  if (activation == 2u) {
    outputs[index] = tanhApprox(sum);
    return;
  }

  outputs[index] = sum;
}
      `.trim();

      const shaderModule = device.createShaderModule({ code: shader });
      const bindGroupLayout = device.createBindGroupLayout({
        entries: [
          {
            binding: 0,
            visibility: 4,
            buffer: { type: 'read-only-storage' },
          },
          {
            binding: 1,
            visibility: 4,
            buffer: { type: 'read-only-storage' },
          },
          {
            binding: 2,
            visibility: 4,
            buffer: { type: 'read-only-storage' },
          },
          {
            binding: 3,
            visibility: 4,
            buffer: { type: 'read-only-storage' },
          },
          {
            binding: 4,
            visibility: 4,
            buffer: { type: 'storage' },
          },
          {
            binding: 5,
            visibility: 4,
            buffer: { type: 'uniform' },
          },
        ],
      });
      const pipelineLayout = device.createPipelineLayout({
        bindGroupLayouts: [bindGroupLayout],
      });
      const pipeline = device.createComputePipeline({
        layout: pipelineLayout,
        compute: {
          module: shaderModule,
          entryPoint: 'main',
        },
      });

      const storageUsage =
        this.resolveBufferUsage('STORAGE', 0x0080) |
        this.resolveBufferUsage('COPY_DST', 0x0008) |
        this.resolveBufferUsage('COPY_SRC', 0x0004);
      const uniformUsage =
        this.resolveBufferUsage('UNIFORM', 0x0040) |
        this.resolveBufferUsage('COPY_DST', 0x0008);
      const readUsage =
        this.resolveBufferUsage('MAP_READ', 0x0001) |
        this.resolveBufferUsage('COPY_DST', 0x0008);

      const weightsBuffer = device.createBuffer({
        size: this.gpuWeights.byteLength,
        usage: storageUsage,
      });
      const biasesBuffer = device.createBuffer({
        size: this.gpuBiases.byteLength,
        usage: storageUsage,
      });
      const inputBuffer = device.createBuffer({
        size: inputs.byteLength,
        usage: storageUsage,
      });
      const activationBuffer = device.createBuffer({
        size: this.gpuActivations.byteLength,
        usage: storageUsage,
      });
      const outputBuffer = device.createBuffer({
        size: totalSize * Float32Array.BYTES_PER_ELEMENT,
        usage: storageUsage,
      });
      const metaBuffer = device.createBuffer({
        size: Uint32Array.BYTES_PER_ELEMENT * 4,
        usage: uniformUsage,
      });
      const readbackBuffer = device.createBuffer({
        size: totalSize * Float32Array.BYTES_PER_ELEMENT,
        usage: readUsage,
      });

      device.queue.writeBuffer(weightsBuffer, 0, this.gpuWeights);
      device.queue.writeBuffer(biasesBuffer, 0, this.gpuBiases);
      device.queue.writeBuffer(inputBuffer, 0, inputs);
      device.queue.writeBuffer(activationBuffer, 0, this.gpuActivations);
      device.queue.writeBuffer(
        metaBuffer,
        0,
        new Uint32Array([this.networkSize, totalSize, 0, 0])
      );

      const bindGroup = device.createBindGroup({
        layout: bindGroupLayout,
        entries: [
          { binding: 0, resource: { buffer: weightsBuffer } },
          { binding: 1, resource: { buffer: biasesBuffer } },
          { binding: 2, resource: { buffer: inputBuffer } },
          { binding: 3, resource: { buffer: activationBuffer } },
          { binding: 4, resource: { buffer: outputBuffer } },
          { binding: 5, resource: { buffer: metaBuffer } },
        ],
      });

      const encoder = device.createCommandEncoder();
      const pass = encoder.beginComputePass?.();
      pass?.setPipeline?.(pipeline);
      pass?.setBindGroup?.(0, bindGroup);
      pass?.dispatchWorkgroups?.(Math.max(1, Math.ceil(totalSize / 64)));
      pass?.end?.();
      encoder.copyBufferToBuffer?.(
        outputBuffer,
        0,
        readbackBuffer,
        0,
        totalSize * Float32Array.BYTES_PER_ELEMENT
      );
      device.queue.submit([encoder.finish?.()]);

      await readbackBuffer.mapAsync?.(this.resolveMapModeRead());
      const mappedRange = readbackBuffer.getMappedRange?.();
      if (!mappedRange) {
        readbackBuffer.unmap?.();
        return null;
      }

      const output = new Float32Array(mappedRange.slice(0));
      readbackBuffer.unmap?.();
      weightsBuffer.destroy?.();
      biasesBuffer.destroy?.();
      inputBuffer.destroy?.();
      activationBuffer.destroy?.();
      outputBuffer.destroy?.();
      metaBuffer.destroy?.();
      readbackBuffer.destroy?.();
      return output;
    } catch {
      return null;
    }
  }
}

export class WebNNEngine extends GPUEngine {
  context: unknown = null;
  builder: unknown = null;
  graph: unknown = null;
  isReady = false;

  override async init(): Promise<void> {
    await super.init();
    const globalNavigator =
      typeof navigator === 'undefined'
        ? undefined
        : (navigator as Navigator & { ml?: unknown });
    this.isReady = Boolean(globalNavigator?.ml);
  }

  async prepareModel(
    size: number,
    weights: Float32Array,
    biases: Float32Array,
    batchSize = 1
  ): Promise<void> {
    this.prepareBuffers(size, weights, biases, batchSize);
  }
}

const verifiedTopologies = new Map<string, Promise<void>>();

async function verifyTopology(topologySource: string): Promise<void> {
  const cached = verifiedTopologies.get(topologySource);
  if (cached) {
    await cached;
    return;
  }

  const pending = (async () => {
    const result = await checkGgProgram(lowerUfcsSource(topologySource), {
      defaults: {
        maxDepth: 64,
        maxBeta1Exclusive: 512,
      },
    });

    if (!result.ok) {
      const message = result.violations
        .map((violation: any) => violation.message)
        .join('; ');
      throw new Error(`Invalid .gg topology for neural engine: ${message}`);
    }
  })();

  verifiedTopologies.set(topologySource, pending);
  try {
    await pending;
  } catch (error) {
    verifiedTopologies.delete(topologySource);
    throw error;
  }
}

export class NeuralEngine {
  gpu: GPUEngine;
  webgpu: WebGPUEngine;
  npu: WebNNEngine;
  neuronRepo: NeuronRepository;
  synapseRepo: SynapseRepository;
  translator: Translator;
  activeBackend: 'gpu' | 'npu' = 'gpu';
  adapterTrainingConfig: AdapterTrainingConfig = {
    rank: 8,
    basePrecision: 'int8',
    adapterPrecision: 'fp16',
    microBatchSize: 16,
    idleFlushMs: 45_000,
  };

  private readonly store: NeuralStore;

  private neurons: Neuron[] = [];
  private synapses: Synapse[] = [];
  private topologyProgram: GgProgram | null = null;
  private topologySource: string;
  private readonly customTopologyProvided: boolean;

  constructor(topologySource?: string) {
    this.store = {
      neurons: new Map(),
      synapses: new Map(),
      semanticDescriptions: new Map(),
    };
    this.customTopologyProvided = topologySource !== undefined;
    this.topologySource = topologySource ?? TOPIC_DOMAIN_TRANSFORMER_TOPOLOGY;
    this.gpu = new GPUEngine();
    this.webgpu = new WebGPUEngine();
    this.npu = new WebNNEngine();
    this.neuronRepo = new NeuronRepository(this.store);
    this.synapseRepo = new SynapseRepository(this.store);
    this.translator = new Translator();
  }

  async init(): Promise<NeuralGraphData> {
    if (!this.customTopologyProvided) {
      this.topologySource = await getTopicDomainTransformerTopologySource();
    }

    await this.loadTopology(this.topologySource, { preserveWeights: false });
    return this.getGraphData();
  }

  async loadTopology(
    topologySource: string,
    options: LoadTopologyOptions = {}
  ): Promise<NeuralGraphData> {
    this.topologySource = topologySource;
    await verifyTopology(this.topologySource);
    this.topologyProgram = parseGgProgram(lowerUfcsSource(this.topologySource));

    await this.gpu.init();
    this.gpu.batchSize = Math.max(1, this.adapterTrainingConfig.microBatchSize);

    await this.webgpu.init();
    this.webgpu.batchSize = Math.max(
      1,
      this.adapterTrainingConfig.microBatchSize
    );

    await this.npu.init();
    if (this.npu.isReady) {
      this.activeBackend = 'npu';
    }

    if (!options.preserveWeights) {
      await this.seedGraphFromTopology();
    }

    await this.compile();
    return this.getGraphData();
  }

  async loadTopologyFile(
    topologyFilePath: string,
    options: LoadTopologyOptions = {}
  ): Promise<NeuralGraphData> {
    const source = await readTopologyFromPath(topologyFilePath);
    return this.loadTopology(source, options);
  }

  getTopologySource(): string {
    return this.topologySource;
  }

  async compile(): Promise<{
    size: number;
    weights: Float32Array;
    biases: Float32Array;
    initialValues: Float32Array;
  }> {
    this.neurons = await this.neuronRepo.getAll();
    this.synapses = await this.synapseRepo.getAll();

    const flattened = this.translator.flatten(this.neurons, this.synapses);
    this.gpu.prepareBuffers(
      flattened.size,
      flattened.weights,
      flattened.biases,
      this.gpu.batchSize
    );
    this.gpu.setActivations(this.neurons.map((neuron) => neuron.activation));

    const targetSize = flattened.size * this.gpu.batchSize;
    this.gpu.prepareTrainingBuffers(
      new Float32Array(targetSize).fill(DEFAULT_TRAINING_IGNORE_TARGET),
      0.001
    );

    if (this.webgpu.isReady) {
      this.webgpu.prepareBuffers(
        flattened.size,
        flattened.weights,
        flattened.biases,
        this.gpu.batchSize
      );
      this.webgpu.setActivations(
        this.neurons.map((neuron) => neuron.activation)
      );
    }

    if (this.npu.isReady) {
      await this.npu.prepareModel(
        flattened.size,
        flattened.weights,
        flattened.biases,
        this.gpu.batchSize
      );
      this.npu.setActivations(this.neurons.map((neuron) => neuron.activation));
    }

    return flattened;
  }

  async getHeteroFabricBackends(
    options: {
      includeEnvCommandBackends?: boolean;
      backends?: HeteroMoAFabricBackendDescriptor[];
    } = {}
  ): Promise<HeteroMoAFabricBackendDescriptor[]> {
    const flattened = await this.compile();
    const activations = this.neurons.map((neuron) => neuron.activation);
    const batchSize = this.gpu.batchSize;
    const descriptors: HeteroMoAFabricBackendDescriptor[] = [
      {
        id: 'cpu-js',
        label: 'CPU lane',
        kind: 'cpu',
        layer: 'cpu',
        priority: 1,
        detail: 'Deterministic in-process CPU execution',
        execute: async (input) => {
          const engine = new GPUEngine();
          await engine.init();
          engine.prepareBuffers(
            flattened.size,
            flattened.weights,
            flattened.biases,
            batchSize
          );
          engine.setActivations(activations);
          return {
            value: await engine.runTick(
              this.toFloat32Input(input, flattened.size * batchSize)
            ),
          };
        },
      },
    ];

    if (this.webgpu.isReady) {
      descriptors.push({
        id: 'webgpu-browser',
        label: 'WebGPU lane',
        kind: 'webgpu',
        layer: 'gpu',
        priority: 3,
        detail: 'Browser WebGPU compute shader execution',
        execute: async (input) => {
          const engine = new WebGPUEngine();
          await engine.init();
          engine.prepareBuffers(
            flattened.size,
            flattened.weights,
            flattened.biases,
            batchSize
          );
          engine.setActivations(activations);
          return {
            value: await engine.runTick(
              this.toFloat32Input(input, flattened.size * batchSize)
            ),
          };
        },
      });
    }

    if (this.npu.isReady) {
      descriptors.push({
        id: 'webnn-browser',
        label: 'WebNN lane',
        kind: 'webnn',
        layer: 'npu',
        priority: 4,
        detail: 'Browser WebNN/NPU execution when available',
        execute: async (input) => {
          const engine = new WebNNEngine();
          await engine.init();
          if (!engine.isReady) {
            throw new Error('WebNN backend is unavailable');
          }
          await engine.prepareModel(
            flattened.size,
            flattened.weights,
            flattened.biases,
            batchSize
          );
          engine.setActivations(activations);
          return {
            value: await engine.runTick(
              this.toFloat32Input(input, flattened.size * batchSize)
            ),
          };
        },
      });
    }

    if (options.includeEnvCommandBackends !== false) {
      descriptors.push(...createCommandFabricBackendsFromEnv());
    }

    if (options.backends) {
      descriptors.push(...options.backends);
    }

    return descriptors;
  }

  async runHeteroMoAFabric(
    inputs: Float32Array,
    options: NeuralHeteroFabricOptions = {}
  ): Promise<HeteroMoAFabricRunResult> {
    const backendOptions: {
      includeEnvCommandBackends?: boolean;
      backends?: HeteroMoAFabricBackendDescriptor[];
    } = {};
    if (options.includeEnvCommandBackends !== undefined) {
      backendOptions.includeEnvCommandBackends =
        options.includeEnvCommandBackends;
    }
    if (options.backends !== undefined) {
      backendOptions.backends = options.backends;
    }
    const backends = await this.getHeteroFabricBackends(backendOptions);

    const runOptions = {
      ...(options.fabricKey !== undefined ? { fabricKey: options.fabricKey } : {}),
      ...('witness' in options && options.witness !== undefined
        ? { witness: options.witness }
        : {}),
      ...('plan' in options && options.plan !== undefined
        ? { plan: options.plan }
        : {}),
      ...(options.inputDocument !== undefined
        ? { inputDocument: options.inputDocument }
        : {}),
      ...(options.communityMemory !== undefined
        ? { communityMemory: options.communityMemory }
        : {}),
      ...(options.compareOutputs !== undefined
        ? { compareOutputs: options.compareOutputs }
        : {}),
      ...(options.isSufficient !== undefined
        ? { isSufficient: options.isSufficient }
        : {}),
      ...(options.streamId !== undefined ? { streamId: options.streamId } : {}),
      ...(options.sequence !== undefined ? { sequence: options.sequence } : {}),
      ...(options.nowMs !== undefined ? { nowMs: options.nowMs } : {}),
      ...(options.wallClockMs !== undefined
        ? { wallClockMs: options.wallClockMs }
        : {}),
    } satisfies HeteroMoAFabricRunOptions;
    return runHeteroMoAFabric(backends, inputs, runOptions);
  }

  private toFloat32Input(
    input: unknown,
    expectedLength?: number
  ): Float32Array {
    const resize = (candidate: Float32Array): Float32Array => {
      if (
        expectedLength === undefined ||
        expectedLength <= 0 ||
        candidate.length === expectedLength
      ) {
        return candidate;
      }
      const resized = new Float32Array(expectedLength);
      resized.set(candidate.slice(0, expectedLength));
      return resized;
    };

    if (input instanceof Float32Array) {
      return resize(input);
    }
    if (ArrayBuffer.isView(input)) {
      const view = input as ArrayBufferView;
      return resize(
        new Float32Array(
          view.buffer.slice(view.byteOffset, view.byteOffset + view.byteLength)
        )
      );
    }
    if (Array.isArray(input)) {
      return resize(new Float32Array(input.map((value) => Number(value) || 0)));
    }
    throw new Error(
      'Neural hetero fabric inputs must be Float32Array-compatible'
    );
  }

  async runTick(inputs: Float32Array): Promise<Float32Array> {
    if (this.activeBackend === 'npu' && this.npu.isReady) {
      return this.npu.runTick(inputs);
    }
    return this.gpu.runTick(inputs);
  }

  async forward(inputs: Float32Array): Promise<Float32Array> {
    return this.runTick(inputs);
  }

  async train(
    inputs: Float32Array,
    targets: Float32Array
  ): Promise<Float32Array> {
    return this.gpu.train(inputs, targets);
  }

  async deployToCloud(): Promise<NeuralGraphData> {
    for (let index = 0; index < this.neurons.length; index++) {
      if (index % 5 === 0) {
        const neuron = this.neurons[index];
        if (neuron === undefined) {
          continue;
        }
        await this.neuronRepo.create({
          ...neuron,
          type: 'cloud',
        });
      }
    }

    this.neurons = await this.neuronRepo.getAll();
    return this.getGraphData();
  }

  setAdapterTrainingConfig(config: Partial<AdapterTrainingConfig>): void {
    this.adapterTrainingConfig = {
      ...this.adapterTrainingConfig,
      ...config,
    };
    this.gpu.batchSize = Math.max(1, this.adapterTrainingConfig.microBatchSize);
  }

  getAdapterTrainingConfig(): AdapterTrainingConfig {
    return { ...this.adapterTrainingConfig };
  }

  getMemoryStats(): { totalNodes: number; totalConnections: number } {
    return {
      totalNodes: this.neurons.length,
      totalConnections: this.synapses.length,
    };
  }

  getGraphData(): NeuralGraphData {
    const neuronToIndex = new Map<string, number>();
    this.neurons.forEach((neuron, index) => {
      neuronToIndex.set(neuron.id, index);
    });

    return {
      nodeCount: this.neurons.length,
      nodes: this.neurons.map((neuron, index) => ({
        id: neuron.id,
        index,
        type: neuron.type,
      })),
      edges: this.synapses.map((synapse) => ({
        id: synapse.id,
        source: neuronToIndex.get(synapse.from_id) ?? 0,
        target: neuronToIndex.get(synapse.to_id) ?? 0,
        weight: synapse.weight,
      })),
    };
  }

  async deleteSynapse(id: string): Promise<NeuralGraphData> {
    await this.synapseRepo.delete(id);
    this.synapses = await this.synapseRepo.getAll();
    await this.compile();
    return this.getGraphData();
  }

  exportGraph(): { version: string; neurons: Neuron[]; synapses: Synapse[] } {
    return {
      version: 'gnosis-neural-gg-1.0',
      neurons: [...this.neurons],
      synapses: [...this.synapses],
    };
  }

  async importGraph(data: {
    neurons: Neuron[];
    synapses: Synapse[];
  }): Promise<NeuralGraphData> {
    if (!Array.isArray(data.neurons) || !Array.isArray(data.synapses)) {
      throw new Error('Invalid graph data');
    }

    this.store.neurons.clear();
    this.store.synapses.clear();

    for (const neuron of data.neurons) {
      await this.neuronRepo.create(neuron);
    }

    for (const synapse of data.synapses) {
      await this.synapseRepo.create(synapse);
    }

    this.neurons = await this.neuronRepo.getAll();
    this.synapses = await this.synapseRepo.getAll();
    await this.compile();
    return this.getGraphData();
  }

  private async seedGraphFromTopology(): Promise<void> {
    this.store.neurons.clear();
    this.store.synapses.clear();

    const program =
      this.topologyProgram ??
      parseGgProgram(lowerUfcsSource(this.topologySource));
    const graph = buildGraphFromGg(program);

    for (const neuron of graph.neurons) {
      await this.neuronRepo.create(neuron);
    }

    for (const synapse of graph.synapses) {
      await this.synapseRepo.create(synapse);
    }

    this.neurons = graph.neurons;
    this.synapses = graph.synapses;
  }
}

/**
 * Handles the gnosis init workflow.
 */
export async function init(): Promise<NeuralEngine> {
  const engine = new NeuralEngine();
  await engine.init();
  return engine;
}
