// @a0n/gnosis/src/option-result.ts
/**
 * option-result.ts -- First-class error values for gnosis topologies
 *
 * Option standardizes absence. Result standardizes success-or-failure.
 * Failures move through the graph as data, not ambient exceptions.
 *
 * Roadmap item #3: depends on ADTs.
 */

import {
  type Variant,
  variant,
  type SumTypeDefinition,
  BUILTIN_SUM_TYPES,
  matchExhaustive,
} from './adt';

// ============================================================================
// Option<T> = Some(T) | None
// ============================================================================

export type GnosisOption<T> = Variant<T>;

export function some<T>(value: T): GnosisOption<T> {
  return variant('some', value);
}

export function none<T>(): GnosisOption<T> {
  return variant('none', undefined as T);
}

export function isSome<T>(opt: GnosisOption<T>): boolean {
  return opt.tag === 'some';
}

export function isNone<T>(opt: GnosisOption<T>): boolean {
  return opt.tag === 'none';
}

export function unwrapOption<T>(opt: GnosisOption<T>): T {
  if (opt.tag === 'none') {
    throw new Error('Attempted to unwrap None');
  }
  return opt.value;
}

export function unwrapOptionOr<T>(opt: GnosisOption<T>, defaultValue: T): T {
  return opt.tag === 'some' ? opt.value : defaultValue;
}

export function mapOption<T, U>(
  opt: GnosisOption<T>,
  f: (v: T) => U
): GnosisOption<U> {
  return opt.tag === 'some' ? some(f(opt.value)) : none<U>();
}

export function flatMapOption<T, U>(
  opt: GnosisOption<T>,
  f: (v: T) => GnosisOption<U>
): GnosisOption<U> {
  return opt.tag === 'some' ? f(opt.value) : none<U>();
}

export function filterOption<T>(
  opt: GnosisOption<T>,
  predicate: (v: T) => boolean
): GnosisOption<T> {
  return opt.tag === 'some' && predicate(opt.value) ? opt : none<T>();
}

// ============================================================================
// Result<T, E> = Ok(T) | Err(E)
// ============================================================================

export type GnosisResult<T, E = Error> = Variant<T | E>;

export function ok<T, E = Error>(value: T): GnosisResult<T, E> {
  return variant('ok', value);
}

export function err<T, E = Error>(error: E): GnosisResult<T, E> {
  return variant('err', error);
}

export function isOk<T, E>(result: GnosisResult<T, E>): boolean {
  return result.tag === 'ok';
}

export function isErr<T, E>(result: GnosisResult<T, E>): boolean {
  return result.tag === 'err';
}

export function unwrapResult<T, E>(result: GnosisResult<T, E>): T {
  if (result.tag === 'err') {
    throw new Error(`Attempted to unwrap Err: ${result.value}`);
  }
  return result.value as T;
}

export function unwrapErr<T, E>(result: GnosisResult<T, E>): E {
  if (result.tag === 'ok') {
    throw new Error(`Attempted to unwrapErr on Ok`);
  }
  return result.value as E;
}

export function unwrapResultOr<T, E>(
  result: GnosisResult<T, E>,
  defaultValue: T
): T {
  return result.tag === 'ok' ? (result.value as T) : defaultValue;
}

export function mapResult<T, U, E>(
  result: GnosisResult<T, E>,
  f: (v: T) => U
): GnosisResult<U, E> {
  return result.tag === 'ok'
    ? ok(f(result.value as T))
    : err(result.value as E);
}

export function mapErr<T, E, F>(
  result: GnosisResult<T, E>,
  f: (e: E) => F
): GnosisResult<T, F> {
  return result.tag === 'err'
    ? err(f(result.value as E))
    : ok(result.value as T);
}

export function flatMapResult<T, U, E>(
  result: GnosisResult<T, E>,
  f: (v: T) => GnosisResult<U, E>
): GnosisResult<U, E> {
  return result.tag === 'ok' ? f(result.value as T) : err(result.value as E);
}

// ============================================================================
// Result from exception -- capture thrown errors as Result
// ============================================================================

export function tryCatch<T>(f: () => T): GnosisResult<T, Error> {
  try {
    return ok(f());
  } catch (e) {
    return err(e instanceof Error ? e : new Error(String(e)));
  }
}

export async function tryCatchAsync<T>(
  f: () => Promise<T>
): Promise<GnosisResult<T, Error>> {
  try {
    return ok(await f());
  } catch (e) {
    return err(e instanceof Error ? e : new Error(String(e)));
  }
}

// ============================================================================
// Combinators -- compose Options and Results through topology flow
// ============================================================================

/** Transpose Option<Result<T,E>> → Result<Option<T>,E> */
export function transposeOptionResult<T, E>(
  opt: GnosisOption<GnosisResult<T, E>>
): GnosisResult<GnosisOption<T>, E> {
  if (opt.tag === 'none') return ok(none<T>());
  const inner = opt.value as GnosisResult<T, E>;
  return inner.tag === 'ok'
    ? ok(some(inner.value as T))
    : err(inner.value as E);
}

/** Collect an array of Results into a Result of array (short-circuit on first Err) */
export function collectResults<T, E>(
  results: GnosisResult<T, E>[]
): GnosisResult<T[], E> {
  const values: T[] = [];
  for (const r of results) {
    if (r.tag === 'err') return err(r.value as E);
    values.push(r.value as T);
  }
  return ok(values);
}

/** Partition Results into [oks, errs] */
export function partitionResults<T, E>(
  results: GnosisResult<T, E>[]
): { oks: T[]; errs: E[] } {
  const oks: T[] = [];
  const errs: E[] = [];
  for (const r of results) {
    if (r.tag === 'ok') oks.push(r.value as T);
    else errs.push(r.value as E);
  }
  return { oks, errs };
}

/** Convert Option to Result (None → Err with provided error) */
export function optionToResult<T, E>(
  opt: GnosisOption<T>,
  error: E
): GnosisResult<T, E> {
  return opt.tag === 'some' ? ok(opt.value) : err(error);
}

/** Convert Result to Option (Err → None) */
export function resultToOption<T, E>(
  result: GnosisResult<T, E>
): GnosisOption<T> {
  return result.tag === 'ok' ? some(result.value as T) : none<T>();
}

// ============================================================================
// Topology integration -- Result/Option as node payload protocol
// ============================================================================

/**
 * Wrap a handler function to return Result instead of throwing.
 * This is the migration path for existing handlers.
 */
export function resultHandler<P, T>(
  handler: (payload: P, props: Record<string, string>) => Promise<T>
): (
  payload: P,
  props: Record<string, string>
) => Promise<GnosisResult<T, Error>> {
  return async (payload, props) => tryCatchAsync(() => handler(payload, props));
}

/**
 * Unwrap a Result payload, routing Ok to the next node and Err to VENT.
 * Returns { value, shouldVent } for the engine to route.
 */
export function routeResult<T, E>(
  result: GnosisResult<T, E>
): { value: unknown; shouldVent: boolean; tag: string } {
  return {
    value: result.value,
    shouldVent: result.tag === 'err',
    tag: result.tag,
  };
}
