// @a0n/gnosis/src/organism-repair.ts
/**
 * Organism Repair — auto-inject VENT paths into deaf branching functions.
 *
 * The scanner identifies functions with FORK/FOLD but no VENT path
 * (ETHICS_MISSING_VENT_PATH). These functions branch but can never reject —
 * they're deaf to their own failure modes.
 *
 * This module generates concrete code patches that add try/catch error
 * handling to the top deaf functions, ordered by the God Formula weight.
 * Each patch converts a deaf function into an active one by adding a
 * rejection channel (VENT path).
 *
 * Two modes:
 *   --dry-run: show what would change, project health trajectory (default)
 *   --fix: apply the patches, then re-scan to verify
 *
 * This is the "integration" repair path from GallWaspTestimony:
 * name the wound, convert it into a bearing structure, beta1 restored.
 *
 * In the AgileGnosisTopology model, each --fix invocation is a sprint:
 *   - AgileFailureIntegration: deaf functions are delta-error fuel
 *   - ReversibilityOperator: try/catch injection converts error into velocity
 *   - PisotTriggeredSprint: fires when sigma degrades, not on a calendar
 *   - thm_waterfall_causes_terminal_collapse: NOT fixing = guaranteed collapse
 *
 * The loop terminates at TopologicalSaturation (EschatonTopology.lean):
 * when all deaf functions have VENT paths, there is no more delta-error
 * to process. GraphSabbath — perfection stops sequence.
 *
 * Uses TypeScript's own parser for precise function boundary detection.
 * The scanner tells us WHICH function (name + line). The TS parser tells
 * us WHERE it starts and ends — exact character offsets, no heuristics.
 *
 * @module @a0n/gnosis/organism-repair
 */

import { readFileSync, writeFileSync } from 'fs';
import { relative } from 'path';
import type { PolyglotAnalysisResult, PolyglotFunctionAnalysis } from './polyglot-bridge';
import * as ts from 'typescript';

// ═══════════════════════════════════════════════════════════════════════
// Types
// ═══════════════════════════════════════════════════════════════════════

/**
 * A concrete repair patch for a single function.
 */
export interface RepairPatch {
  /** Source file path. */
  filePath: string;
  /** Function name. */
  functionName: string;
  /** Line number where the function starts. */
  startLine: number;
  /** Line number where the function ends. */
  endLine: number;
  /** The diagnostic that triggered the repair. */
  diagnostic: string;
  /** What the patch does, in plain English. */
  description: string;
  /** The original function source. */
  originalSource: string;
  /** The patched function source. */
  patchedSource: string;
  /** God Formula weight (higher = fix first). */
  godWeight: number;
  /** Buley complexity measure. */
  buleyMeasure: number;
}

/**
 * A repair plan for a project.
 */
export interface RepairPlan {
  /** Project path. */
  projectPath: string;
  /** Total deaf branching functions found. */
  totalDeafBranching: number;
  /** Number of patches generated (capped by limit). */
  patchCount: number;
  /** The patches, ordered by Buley complexity descending. */
  patches: RepairPatch[];
  /** Estimated sigma reduction if all patches applied. */
  estimatedSigmaReduction: number;
}

// ═══════════════════════════════════════════════════════════════════════
// TS Parser — find exact function boundaries
// ═══════════════════════════════════════════════════════════════════════

/** Cached parsed source files to avoid re-parsing. */
const parsedFileCache = new Map<string, ts.SourceFile>();

/**
 * Parse a TypeScript source file (cached).
 */
function getParsedFile(filePath: string, source: string): ts.SourceFile {
  let sf = parsedFileCache.get(filePath);
  if (!sf) {
    sf = ts.createSourceFile(
      filePath,
      source,
      ts.ScriptTarget.Latest,
      /* setParentNodes */ true,
    );
    parsedFileCache.set(filePath, sf);
  }
  return sf;
}

/**
 * Find a function/method declaration by name and approximate line number.
 * Returns the exact character offsets of the function and its body block.
 *
 * Handles:
 *   - function declarations: function foo() { ... }
 *   - exported function declarations: export function foo() { ... }
 *   - arrow functions: const foo = () => { ... }
 *   - method declarations: foo() { ... }
 */
function findFunctionInAST(
  sf: ts.SourceFile,
  functionName: string,
  hintLine: number,
): { funcStart: number; funcEnd: number; bodyStart: number; bodyEnd: number } | null {
  let best: { node: ts.Node; body: ts.Block } | null = null;
  let bestDist = Infinity;

  function visit(node: ts.Node) {
    let name: string | undefined;
    let body: ts.Block | undefined;

    // function foo() { ... }
    if (ts.isFunctionDeclaration(node) && node.name && node.body) {
      name = node.name.text;
      body = node.body;
    }
    // class method: foo() { ... }
    else if (ts.isMethodDeclaration(node) && node.body && ts.isIdentifier(node.name)) {
      name = node.name.text;
      body = node.body;
    }
    // const foo = function() { ... } or const foo = () => { ... }
    else if (ts.isVariableDeclaration(node) && ts.isIdentifier(node.name) && node.initializer) {
      const init = node.initializer;
      if (
        (ts.isArrowFunction(init) || ts.isFunctionExpression(init)) &&
        init.body && ts.isBlock(init.body)
      ) {
        name = node.name.text;
        body = init.body;
      }
    }

    if (name === functionName && body) {
      const line = sf.getLineAndCharacterOfPosition(node.getStart(sf)).line + 1;
      const dist = Math.abs(line - hintLine);
      if (dist < bestDist) {
        bestDist = dist;
        best = { node, body };
      }
    }

    ts.forEachChild(node, visit);
  }

  visit(sf);
  if (best === null) return null;
  const found = best as { node: ts.Node; body: ts.Block };

  // For variable declarations, we want the full statement (including const/export)
  let funcNode = found.node;
  if (ts.isVariableDeclaration(funcNode)) {
    // Walk up to the VariableStatement
    if (funcNode.parent && ts.isVariableDeclarationList(funcNode.parent)) {
      if (funcNode.parent.parent && ts.isVariableStatement(funcNode.parent.parent)) {
        funcNode = funcNode.parent.parent;
      }
    }
  }

  return {
    funcStart: funcNode.getStart(sf),
    funcEnd: funcNode.getEnd(),
    bodyStart: found.body.getStart(sf) + 1, // character after '{'
    bodyEnd: found.body.getEnd() - 1,       // character before '}'
  };
}

// ═══════════════════════════════════════════════════════════════════════
// Patch Generation
// ═══════════════════════════════════════════════════════════════════════

/**
 * Generate a VENT patch for a single function using the TS parser.
 *
 * The scanner tells us the function name and line. The TS parser gives
 * exact boundaries. We splice a try/catch around the body.
 */
function generateVentPatch(
  func: PolyglotFunctionAnalysis,
  fileSource: string,
  filePath: string,
): RepairPatch | null {
  // Get the line hint from the scanner's source map
  const entrySpan = func.sourceMap.get('entry_0');
  if (!entrySpan) return null;
  const hintLine = entrySpan.start_line;

  // Use the TS parser to find exact function boundaries
  const sf = getParsedFile(filePath, fileSource);
  const bounds = findFunctionInAST(sf, func.functionName, hintLine);
  if (!bounds) return null;

  // Extract original source
  const originalSource = fileSource.slice(bounds.funcStart, bounds.funcEnd);

  // Safety: don't double-wrap functions that already have try/catch
  if (/\btry\s*\{/.test(originalSource)) return null;

  // Extract the body (between { and })
  const body = fileSource.slice(bounds.bodyStart, bounds.bodyEnd);
  if (body.trim().length === 0) return null;

  // Detect body indentation from the first non-empty line
  let indent = '  ';
  for (const line of body.split('\n')) {
    const m = line.match(/^(\s+)\S/);
    if (m) { indent = m[1]; break; }
  }

  // Re-indent body by +2 spaces for try block
  const indentedBody = body.split('\n').map(line => {
    if (line.trim() === '') return line;
    return '  ' + line;
  }).join('\n');

  // Escape function name for template literal
  const safeName = func.functionName.replace(/[`$\\]/g, '\\$&');

  // Build patched source:
  // Everything before the body + try { indented body } catch { ... } + closing
  const beforeBody = fileSource.slice(bounds.funcStart, bounds.bodyStart);
  const afterBody = fileSource.slice(bounds.bodyEnd, bounds.funcEnd);

  const patchedSource =
    beforeBody + '\n' +
    indent + 'try {' +
    indentedBody + '\n' +
    indent + '} catch (error) {\n' +
    indent + '  throw new Error(`[' + safeName + '] ${error instanceof Error ? error.message : String(error)}`);\n' +
    indent + '}' + '\n' +
    afterBody;

  // Compute God Formula weight
  const beta1 = func.bettyResult.b1;
  const deficit = func.bettyResult.deficit?.totalDeficit ?? 0;
  const totalRounds = beta1 + deficit;
  const godWeight = totalRounds - Math.min(deficit, totalRounds) + 1;

  const startLine = sf.getLineAndCharacterOfPosition(bounds.funcStart).line + 1;
  const endLine = sf.getLineAndCharacterOfPosition(bounds.funcEnd).line + 1;

  return {
    filePath,
    functionName: func.functionName,
    startLine,
    endLine,
    diagnostic: 'ETHICS_MISSING_VENT_PATH',
    description:
      `Wrap ${func.functionName} body in try/catch to add a rejection channel (VENT path). ` +
      `This converts a deaf branching function into an active one.`,
    originalSource,
    patchedSource,
    godWeight,
    buleyMeasure: func.bettyResult.buleyMeasure,
  };
}

// ═══════════════════════════════════════════════════════════════════════
// Repair Plan Generation
// ═══════════════════════════════════════════════════════════════════════

/**
 * Generate a repair plan for a set of analyzed files.
 *
 * Finds all deaf branching functions (have FORK/FOLD but no VENT,
 * flagged with ETHICS_MISSING_VENT_PATH) and generates try/catch
 * patches for the top N by Buley complexity.
 *
 * @param fileResults — polyglot analysis results per file
 * @param projectPath — root path for relative display
 * @param limit — max patches to generate (default 10)
 */
export function generateRepairPlan(
  fileResults: Array<{ filePath: string; result: PolyglotAnalysisResult }>,
  projectPath: string,
  limit: number = 10,
): RepairPlan {
  // Collect all deaf branching functions with ETHICS_MISSING_VENT_PATH
  const candidates: Array<{
    func: PolyglotFunctionAnalysis;
    filePath: string;
    buleyMeasure: number;
  }> = [];

  for (const { filePath, result } of fileResults) {
    for (const func of result.functions) {
      const hasMissingVent = func.bettyResult.diagnostics.some(
        (d) => d.code === 'ETHICS_MISSING_VENT_PATH',
      );
      // Only patch functions that are deaf AND have branching
      // (pure linear functions don't need VENT paths)
      const edges = func.bettyResult.ast?.edges ?? [];
      const hasFork = edges.some((e) => e.type === 'FORK');

      if (hasMissingVent && hasFork) {
        candidates.push({
          func,
          filePath,
          buleyMeasure: func.bettyResult.buleyMeasure,
        });
      }
    }
  }

  // Sort by Buley measure descending — most complex first
  candidates.sort((a, b) => b.buleyMeasure - a.buleyMeasure);

  // Generate patches for the top N candidates
  // Try up to 2× the limit to account for functions we can't patch
  const patches: RepairPatch[] = [];
  const fileSources = new Map<string, string>();

  // Clear the parse cache for fresh parsing
  parsedFileCache.clear();

  for (const candidate of candidates.slice(0, limit * 2)) {
    // Read and cache file source
    if (!fileSources.has(candidate.filePath)) {
      try {
        fileSources.set(
          candidate.filePath,
          readFileSync(candidate.filePath, 'utf8'),
        );
      } catch {
        continue;
      }
    }
    const fileSource = fileSources.get(candidate.filePath)!;

    const patch = generateVentPatch(candidate.func, fileSource, candidate.filePath);
    if (patch) {
      patches.push(patch);
      if (patches.length >= limit) break;
    }
  }

  // Estimate sigma reduction:
  // Each patched function goes from deaf to active (beta1 += 1 effective).
  const totalFuncs = fileResults.reduce((s, f) => s + f.result.functions.length, 0);
  const estimatedSigmaReduction = totalFuncs > 0
    ? patches.length / totalFuncs
    : 0;

  return {
    projectPath,
    totalDeafBranching: candidates.length,
    patchCount: patches.length,
    patches,
    estimatedSigmaReduction,
  };
}

// ═══════════════════════════════════════════════════════════════════════
// Apply Patches
// ═══════════════════════════════════════════════════════════════════════

/**
 * Apply repair patches to files.
 *
 * For each patch, reads the file, verifies the original source is still
 * present, replaces it with the patched source, and writes back.
 *
 * IMPORTANT: patches within the same file must be applied in reverse order
 * (bottom-up) to preserve character offsets. This function handles that.
 *
 * Returns the count of successfully applied patches.
 */
export function applyRepairPlan(plan: RepairPlan): {
  applied: number;
  skipped: number;
  errors: string[];
} {
  let applied = 0;
  let skipped = 0;
  const errors: string[] = [];

  // Group patches by file and sort by startLine descending (bottom-up)
  // so that applying one patch doesn't shift the offsets of later patches.
  const byFile = new Map<string, RepairPatch[]>();
  for (const patch of plan.patches) {
    const list = byFile.get(patch.filePath) ?? [];
    list.push(patch);
    byFile.set(patch.filePath, list);
  }

  for (const [filePath, patches] of byFile) {
    // Sort bottom-up (highest startLine first)
    patches.sort((a, b) => b.startLine - a.startLine);

    let fileSource: string;
    try {
      fileSource = readFileSync(filePath, 'utf8');
    } catch (e: unknown) {
      const msg = e instanceof Error ? e.message : String(e);
      errors.push(`${filePath}: ${msg}`);
      skipped += patches.length;
      continue;
    }

    for (const patch of patches) {
      // Verify the original source is still present
      const idx = fileSource.indexOf(patch.originalSource);
      if (idx === -1) {
        skipped++;
        errors.push(
          `${filePath}:${patch.startLine} ${patch.functionName} — ` +
          `source changed since scan, skipping`,
        );
        continue;
      }

      // Apply the patch
      fileSource =
        fileSource.slice(0, idx) +
        patch.patchedSource +
        fileSource.slice(idx + patch.originalSource.length);
      applied++;
    }

    // Write the file back once with all patches for this file applied
    try {
      writeFileSync(filePath, fileSource, 'utf8');
    } catch (e: unknown) {
      const msg = e instanceof Error ? e.message : String(e);
      errors.push(`${filePath}: write failed: ${msg}`);
    }
  }

  return { applied, skipped, errors };
}
