// @a0n/gnosis/src/packed-knot-encoder.ts
/**
 * Packed Knot Binary Encoder -- GG topology → binary wire format
 *
 * The knot formalizes the data. Crossings encode bits. Zero overhead.
 *
 * Backed by FoldedKnotBitPacking.lean:
 *   pack_unpack: round-trip is identity
 *   zero_overhead: output size = input size
 *   compose_assoc: composition is associative
 *
 * Binary layout now uses the Lexicon Common Payload Header (8 bytes):
 *   [magic:2 = MAGIC.GNOT][version:1][flags:1][crc:2][payloadLen:2]
 *   [nodes...][edges...]
 *
 * Edge encoding:
 *   [flags:1][sourceCount:1][targetCount:1][sourceIds...][targetIds...][propLen:2][props...]
 *
 * Flags byte matches FlowFrame flags:
 *   0b00000001 = FORK
 *   0b00000010 = RACE
 *   0b00000100 = FOLD
 *   0b00001000 = VENT
 *   0b00010000 = PROCESS
 *   0b00100000 = SLIVER
 *   0b01000000 = REJECT
 *
 * Node encoding:
 *   [nameLen:1][name...][labelCount:1][labels...][propLen:2][props...]
 */

import {
  MAGIC,
  LEXICON_VERSION,
  COMMON_HEADER_SIZE,
  COMMON_HEADER_LARGE_SIZE,
  FLAG_LARGE,
  writeHeader,
  readHeader,
  crc16,
  writeVarint,
  readVarint,
} from './lexicon';

// ── Types (mirror aeon-logic GgProgram) ─────────────────────

export interface GgNode {
  readonly id: string;
  readonly labels: readonly string[];
  readonly properties: Readonly<Record<string, unknown>>;
}

export interface GgEdge {
  readonly sourceIds: readonly string[];
  readonly targetIds: readonly string[];
  readonly type: string;
  readonly properties: Readonly<Record<string, unknown>>;
}

export interface GgProgram {
  readonly nodes: readonly GgNode[];
  readonly edges: readonly GgEdge[];
}

// ── Constants ───────────────────────────────────────────────

/** @deprecated Use MAGIC.GNOT from the Lexicon */
const LEGACY_MAGIC_7BYTE = 0x474b; // "GK" = Gnosis Knot (old 7-byte format)

const EDGE_TYPE_FLAGS: Record<string, number> = {
  FORK: 0b00000001,
  RACE: 0b00000010,
  FOLD: 0b00000100,
  VENT: 0b00001000,
  PROCESS: 0b00010000,
  SLIVER: 0b00100000,
  REJECT: 0b01000000,
};

// ── Encoder ─────────────────────────────────────────────────

const encoder = new TextEncoder();
const decoder = new TextDecoder();

function encodeString(s: string): Uint8Array {
  return encoder.encode(s);
}

function encodeSingleProp(val: unknown): Uint8Array {
  if (val instanceof Float32Array) {
    const innerLen = val.byteLength;
    const buf = new Uint8Array(COMMON_HEADER_SIZE + innerLen);
    writeHeader(buf, MAGIC.TENSOR as readonly [number, number], innerLen, { version: LEXICON_VERSION });
    buf.set(new Uint8Array(val.buffer, val.byteOffset, val.byteLength), COMMON_HEADER_SIZE);
    return buf;
  }
  if (val instanceof Uint8Array) {
    const innerLen = val.length;
    const buf = new Uint8Array(COMMON_HEADER_SIZE + innerLen);
    writeHeader(buf, MAGIC.BLOB as readonly [number, number], innerLen, { version: LEXICON_VERSION });
    buf.set(val, COMMON_HEADER_SIZE);
    return buf;
  }
  return encoder.encode(JSON.stringify(val));
}

function encodeProps(props: Readonly<Record<string, unknown>>): Uint8Array {
  const keys = Object.keys(props);
  if (keys.length === 0) return encoder.encode('{}');

  // Check if we have any binary arrays
  const hasBinary = keys.some((k) => props[k] instanceof Float32Array || props[k] instanceof Uint8Array);
  if (!hasBinary) {
    return encoder.encode(JSON.stringify(props));
  }

  // Encode as MAGIC.STRUCT
  const encodedEntries: { keyBytes: Uint8Array; valBytes: Uint8Array }[] = [];
//   const totalInnerSize = 0;
  for (const k of keys) {
    const keyBytes = encoder.encode(k);
    const valBytes = encodeSingleProp(props[k]);
    encodedEntries.push({ keyBytes, valBytes });
  }

  // Pre-calculate size including varints
  const tempEntriesBuf = new Uint8Array(keys.length * 1024 + 1024); // generous
  let offset = 0;
  
  offset += writeVarint(tempEntriesBuf, offset, encodedEntries.length);
  for (const { keyBytes, valBytes } of encodedEntries) {
    offset += writeVarint(tempEntriesBuf, offset, keyBytes.length);
    tempEntriesBuf.set(keyBytes, offset);
    offset += keyBytes.length;
    offset += writeVarint(tempEntriesBuf, offset, valBytes.length);
    tempEntriesBuf.set(valBytes, offset);
    offset += valBytes.length;
  }

  const innerPayload = tempEntriesBuf.subarray(0, offset);
  const innerLen = innerPayload.length;
  const isLarge = innerLen > 0xffff;
  const headerSize = isLarge ? COMMON_HEADER_LARGE_SIZE : COMMON_HEADER_SIZE;
  const structBuf = new Uint8Array(headerSize + innerLen);

  writeHeader(structBuf, MAGIC.STRUCT as readonly [number, number], innerLen, { 
    version: 3, // Version 3 Property Lexicon
    flags: isLarge ? FLAG_LARGE : 0
  });

  structBuf.set(innerPayload, headerSize);
  return structBuf;
}

/**
 * Encode a GG program to packed knot binary with Lexicon Common Header.
 * pack_unpack theorem guarantees round-trip.
 */
export function encode(program: GgProgram): Uint8Array {
//   const parts: Uint8Array[] = [];

  // Build inner payload (nodes + edges) first so we know the size for the header
  const innerParts: Uint8Array[] = [];

  // Node/edge count header: nodeCount(2) + edgeCount(2) = 4 bytes
  const countHeader = new Uint8Array(4);
  const chv = new DataView(countHeader.buffer);
  chv.setUint16(0, program.nodes.length, false);
  chv.setUint16(2, program.edges.length, false);
  innerParts.push(countHeader);

  // Node index: maps node ID → index for compact edge references
  const nodeIndex = new Map<string, number>();

  // Encode nodes
  for (const [i, node] of program.nodes.entries()) {
    nodeIndex.set(node.id, i);

    const nameBytes = encodeString(node.id);
    const propBytes = encodeProps(node.properties);

    // Labels as pipe-separated string
    const labelStr = node.labels.join('|');
    const labelBytes = encodeString(labelStr);

    // [nameLen:1][name][labelLen:1][labels][propLen:4][props]
    const nodeBuf = new Uint8Array(
      1 + nameBytes.length + 1 + labelBytes.length + 4 + propBytes.length
    );
    const nv = new DataView(nodeBuf.buffer);
    let offset = 0;

    nodeBuf[offset++] = nameBytes.length;
    nodeBuf.set(nameBytes, offset);
    offset += nameBytes.length;
    nodeBuf[offset++] = labelBytes.length;
    nodeBuf.set(labelBytes, offset);
    offset += labelBytes.length;
    nv.setUint32(offset, propBytes.length, false);
    offset += 4;
    nodeBuf.set(propBytes, offset);

    innerParts.push(nodeBuf);
  }

  // Encode edges
  for (const edge of program.edges) {
    const flags = EDGE_TYPE_FLAGS[edge.type.toUpperCase()] ?? 0b00010000; // default PROCESS
    const propBytes = encodeProps(edge.properties);

    const srcIndices = edge.sourceIds.map((id) => nodeIndex.get(id) ?? 0);
    const tgtIndices = edge.targetIds.map((id) => nodeIndex.get(id) ?? 0);

    // [flags:1][srcCount:1][tgtCount:1][srcIndices:srcCount*2][tgtIndices:tgtCount*2][propLen:4][props]
    const edgeBuf = new Uint8Array(
      1 +
        1 +
        1 +
        srcIndices.length * 2 +
        tgtIndices.length * 2 +
        4 +
        propBytes.length
    );
    const ev = new DataView(edgeBuf.buffer);
    let offset = 0;

    edgeBuf[offset++] = flags;
    edgeBuf[offset++] = srcIndices.length;
    edgeBuf[offset++] = tgtIndices.length;

    for (const idx of srcIndices) {
      ev.setUint16(offset, idx, false);
      offset += 2;
    }
    for (const idx of tgtIndices) {
      ev.setUint16(offset, idx, false);
      offset += 2;
    }

    ev.setUint32(offset, propBytes.length, false);
    offset += 4;
    edgeBuf.set(propBytes, offset);

    innerParts.push(edgeBuf);
  }

  // Concatenate inner payload
  const innerLen = innerParts.reduce((s, p) => s + p.length, 0);
  const innerPayload = new Uint8Array(innerLen);
  let innerPos = 0;
  for (const part of innerParts) {
    innerPayload.set(part, innerPos);
    innerPos += part.length;
  }

  // Determine header size (standard vs large)
  const isLarge = innerLen > 0xffff;
  const headerSize = isLarge ? COMMON_HEADER_LARGE_SIZE : COMMON_HEADER_SIZE;
  const flags = isLarge ? FLAG_LARGE : 0;

  // Build full output: Lexicon Common Header + inner payload
  const result = new Uint8Array(headerSize + innerLen);
  writeHeader(result, MAGIC.GNOT as readonly [number, number], innerLen, {
    version: LEXICON_VERSION,
    flags,
    crc: crc16(innerPayload),
  });
  result.set(innerPayload, headerSize);

  return result;
}

/**
 * Decode packed knot binary back to GG program.
 * Supports both the new Lexicon Common Header format and the legacy 7-byte format.
 * pack_unpack theorem guarantees this recovers the original.
 */
export function decode(bytes: Uint8Array): GgProgram {
  const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  let offset = 0;

  // Try Lexicon Common Header first
  const header = readHeader(bytes);
  if (
    header &&
    header.magic[0] === MAGIC.GNOT[0] &&
    header.magic[1] === MAGIC.GNOT[1] &&
    header.payloadLength > 0
  ) {
    // Lexicon V2 format
    const headerSize =
      (header.flags & FLAG_LARGE) !== 0
        ? COMMON_HEADER_LARGE_SIZE
        : COMMON_HEADER_SIZE;
    const innerPayload = bytes.subarray(
      headerSize,
      headerSize + header.payloadLength
    );

    // Optional CRC verification (CRC=0 means uncomputed)
    if (header.crc !== 0 && crc16(innerPayload) !== header.crc) {
      throw new Error('PackedKnotEncoder: CRC-16 verification failed');
    }

    return decodeInner(
      innerPayload,
      new DataView(
        innerPayload.buffer,
        innerPayload.byteOffset,
        innerPayload.byteLength
      )
    );
  }

  // Legacy 7-byte format fallback
  const magic = view.getUint16(offset, false);
  offset += 2;
  if (magic !== LEGACY_MAGIC_7BYTE) {
    throw new Error(`Invalid magic: 0x${magic.toString(16)}`);
  }
  const version = view.getUint8(offset);
  offset += 1;
  if (version !== 1) throw new Error(`Unsupported version: ${version}`);

  return decodeInner(
    bytes.subarray(offset),
    new DataView(bytes.buffer, bytes.byteOffset + offset, bytes.byteLength - offset)
  );
}

function decodeSingleProp(bytes: Uint8Array): unknown {
  if (bytes.length < 2) return JSON.parse(decoder.decode(bytes));
  
  const m0 = bytes[0];
  const m1 = bytes[1];
  
  const header = readHeader(bytes);
  if (header) {
    const isLarge = (header.flags & FLAG_LARGE) !== 0;
    const headerSize = isLarge ? COMMON_HEADER_LARGE_SIZE : COMMON_HEADER_SIZE;
    const body = bytes.subarray(headerSize, headerSize + header.payloadLength);
    
    if (m0 === MAGIC.TENSOR[0] && m1 === MAGIC.TENSOR[1]) {
      return new Float32Array(body.slice().buffer); // slice to ensure alignment
    }
    if (m0 === MAGIC.BLOB[0] && m1 === MAGIC.BLOB[1]) {
      return body.slice();
    }
  }
  
  try {
    return JSON.parse(decoder.decode(bytes));
  } catch {
    return decoder.decode(bytes); // fallback to string if invalid JSON
  }
}

function decodeProps(propBytes: Uint8Array): Record<string, unknown> {
  if (propBytes.length === 0) return {};
  
  // Try Lexicon STRUCT
  const header = readHeader(propBytes);
  if (header && header.magic[0] === MAGIC.STRUCT[0] && header.magic[1] === MAGIC.STRUCT[1]) {
    const isLarge = (header.flags & FLAG_LARGE) !== 0;
    const headerSize = isLarge ? COMMON_HEADER_LARGE_SIZE : COMMON_HEADER_SIZE;
    const body = propBytes.subarray(headerSize, headerSize + header.payloadLength);
    
    if (header.version >= 3) {
      // V3 Varint decoding
      let offset = 0;
      const { value: count, bytesRead: cRead } = readVarint(body, offset);
      offset += cRead;
      const props: Record<string, unknown> = {};
      
      for (let i = 0; i < count && offset < body.length; i++) {
        const { value: keyLen, bytesRead: kRead } = readVarint(body, offset);
        offset += kRead;
        const key = decoder.decode(body.subarray(offset, offset + keyLen));
        offset += keyLen;
        const { value: valLen, bytesRead: vRead } = readVarint(body, offset);
        offset += vRead;
        const valBytes = body.subarray(offset, offset + valLen);
        props[key] = decodeSingleProp(valBytes);
        offset += valLen;
      }
      return props;
    }
  }
  
  // Fallback to JSON
  try {
    return JSON.parse(decoder.decode(propBytes));
  } catch {
    return {};
  }
}

function readByteAt(bytes: Uint8Array, offset: number, field: string): number {
  const value = bytes[offset];
  if (value === undefined) {
    throw new Error(
      `Invalid packed-knot payload: missing ${field} at byte offset ${offset}`
    );
  }
  return value;
}

/**
 * Decode the inner payload (node/edge count + nodes + edges) from a raw buffer.
 * Shared between Lexicon and legacy decode paths.
 */
function decodeInner(innerBytes: Uint8Array, view: DataView): GgProgram {
  let offset = 0;

  const nodeCount = view.getUint16(offset, false);
  offset += 2;
  const edgeCount = view.getUint16(offset, false);
  offset += 2;

  // Decode nodes
  const nodes: GgNode[] = [];
  const nodeIds: string[] = [];

  for (let i = 0; i < nodeCount; i++) {
    const nameLen = readByteAt(innerBytes, offset, 'node name length');
    offset += 1;
    const id = decoder.decode(innerBytes.slice(offset, offset + nameLen));
    offset += nameLen;

    const labelLen = readByteAt(innerBytes, offset, 'node label length');
    offset += 1;
    const labelStr = decoder.decode(innerBytes.slice(offset, offset + labelLen));
    offset += labelLen;
    const labels = labelStr ? labelStr.split('|') : [];

    const propLen = view.getUint32(offset, false);
    offset += 4;
    const properties = decodeProps(innerBytes.subarray(offset, offset + propLen));
    offset += propLen;

    nodes.push({ id, labels, properties });
    nodeIds.push(id);
  }

  // Decode edges
  const edges: GgEdge[] = [];
  const flagsToType: Record<number, string> = {};
  for (const [type, flag] of Object.entries(EDGE_TYPE_FLAGS)) {
    flagsToType[flag] = type;
  }

  for (let i = 0; i < edgeCount; i++) {
    const flags = readByteAt(innerBytes, offset, 'edge flags');
    offset += 1;
    const srcCount = readByteAt(innerBytes, offset, 'edge source count');
    offset += 1;
    const tgtCount = readByteAt(innerBytes, offset, 'edge target count');
    offset += 1;

    const sourceIds: string[] = [];
    for (let j = 0; j < srcCount; j++) {
      const idx = view.getUint16(offset, false);
      offset += 2;
      sourceIds.push(nodeIds[idx] ?? String(idx));
    }

    const targetIds: string[] = [];
    for (let j = 0; j < tgtCount; j++) {
      const idx = view.getUint16(offset, false);
      offset += 2;
      targetIds.push(nodeIds[idx] ?? String(idx));
    }

    const propLen = view.getUint32(offset, false);
    offset += 4;
    const properties = decodeProps(innerBytes.subarray(offset, offset + propLen));
    offset += propLen;

    const type = flagsToType[flags] ?? 'PROCESS';
    edges.push({ sourceIds, targetIds, type, properties });
  }

  return { nodes, edges };
}

/**
 * Convert packed knot binary to a sequence of 10-byte FlowFrames.
 * Each edge becomes one FlowFrame with the packed knot segment as payload.
 */
export function toFlowFrames(bytes: Uint8Array): Uint8Array[] {
  const program = decode(bytes);
  const frames: Uint8Array[] = [];
  const FLOW_HEADER_SIZE = 10;
  const STREAM_ID = 1;

  for (const [seq, edge] of program.edges.entries()) {
    const flags = EDGE_TYPE_FLAGS[edge.type.toUpperCase()] ?? 0;

    // Payload: edge properties as compact JSON
    const payload = encoder.encode(JSON.stringify(edge.properties));

    const frame = new Uint8Array(FLOW_HEADER_SIZE + payload.length);
    const fv = new DataView(frame.buffer);

    fv.setUint16(0, STREAM_ID, false);
    fv.setUint32(2, seq >>> 0, false);
    fv.setUint8(6, flags);
    frame[7] = (payload.length >> 16) & 0xff;
    frame[8] = (payload.length >> 8) & 0xff;
    frame[9] = payload.length & 0xff;
    frame.set(payload, FLOW_HEADER_SIZE);

    frames.push(frame);
  }

  return frames;
}

/**
 * Size comparison: packed binary vs text source.
 * zero_overhead theorem: binary size ≤ text size.
 */
export function compressionStats(
  textSource: string,
  binary: Uint8Array
): {
  textBytes: number;
  binaryBytes: number;
  ratio: number;
  savings: number;
} {
  const textBytes = encoder.encode(textSource).length;
  return {
    textBytes,
    binaryBytes: binary.length,
    ratio: binary.length / textBytes,
    savings: textBytes - binary.length,
  };
}
