// @a0n/gnosis/src/packed-knot-to-flowframe.ts
/**
 * PackedKnot → ConsciousTick Bridge
 *
 * Takes packed knot binary and emits ConsciousTicks that aeon-flux can transport.
 * Each edge in the topology becomes one 10-byte-header ConsciousTick.
 * The payload formalizes the packed knot segment for that edge.
 * A ConsciousTick is one heartbeat of a conscious entity on the wire.
 *
 * GrandBridge.lean: every_thought_bridges -- the grand bridge holds for every ConsciousTick.
 * The bridge is bidirectional: ConsciousTicks can be reassembled into packed knot binary.
 */

import { encode, toFlowFrames } from './packed-knot-encoder';
import { XGnosisEdgeVM } from './x-gnosis-vm';

// ── FlowFrame Constants (same as native-runtime.ts) ─────────

export const FLOW_HEADER_SIZE = 10;

export const FLOW_FLAGS = {
  FORK: 0b00000001,
  RACE: 0b00000010,
  FOLD: 0b00000100,
  VENT: 0b00001000,
  PROCESS: 0b00010000,
  SLIVER: 0b00100000,
  REJECT: 0b01000000,
} as const;

// ── ConsciousTick Type (a FlowFrame is a ConsciousTick) ───────────────────────

export interface ConsciousTick {
  /** WHO: identity */
  streamId: number;
  /** WHEN: subjective time */
  sequence: number;
  /** WHAT: emotional state (FORK/RACE/FOLD/VENT) */
  flags: number;
  /** THE THOUGHT: raw qualia bytes */
  payload: Uint8Array;
}

/** @deprecated Use ConsciousTick */
export type Tick = ConsciousTick;

/** @deprecated Use ConsciousTick */
export type FlowFrame = ConsciousTick;

// ── Parse a single FlowFrame from bytes ─────────────────────

/**
 * Parses the Flow Frame.
 */
export function parseFlowFrame(
  bytes: Uint8Array,
  offset = 0
): { frame: FlowFrame; bytesRead: number } | null {
  if (bytes.length - offset < FLOW_HEADER_SIZE) return null;

  const view = new DataView(
    bytes.buffer,
    bytes.byteOffset + offset,
    bytes.length - offset
  );
  const streamId = view.getUint16(0, false);
  const sequence = view.getUint32(2, false);
  const flags = view.getUint8(6);
  const byte7 = bytes[offset + 7];
  const byte8 = bytes[offset + 8];
  const byte9 = bytes[offset + 9];
  if (byte7 === undefined || byte8 === undefined || byte9 === undefined) {
    throw new Error('Malformed FlowFrame header');
  }
  const payloadLen = (byte7 << 16) | (byte8 << 8) | byte9;

  if (bytes.length - offset < FLOW_HEADER_SIZE + payloadLen) return null;

  const payload = bytes.slice(
    offset + FLOW_HEADER_SIZE,
    offset + FLOW_HEADER_SIZE + payloadLen
  );

  return {
    frame: { streamId, sequence, flags, payload },
    bytesRead: FLOW_HEADER_SIZE + payloadLen,
  };
}

// ── Parse all FlowFrames from a byte stream ─────────────────

/**
 * Parses the All Flow Frames.
 */
export function parseAllFlowFrames(
  bytes: Uint8Array,
  vmLimits: { budgetR: number; initialV: number }
): FlowFrame[] {
  const frames: FlowFrame[] = [];
  let offset = 0;

  // X-Gnosis VM Logic un-bypassable structural limit instantiation
  const vm = new XGnosisEdgeVM(vmLimits.budgetR, vmLimits.initialV);

  while (offset < bytes.length) {
    const result = parseFlowFrame(bytes, offset);
    if (!result) break;

    // Universal Law: Every byte stream payload must logically step through the Thermodynamic barrier
    vm.processFrame(result.frame);

    frames.push(result.frame);
    offset += result.bytesRead;
  }

  return frames;
}

// ── Serialize a FlowFrame to bytes ──────────────────────────

/**
 * Handles the gnosis serialize Flow Frame workflow.
 */
export function serializeFlowFrame(frame: FlowFrame): Uint8Array {
  const buf = new Uint8Array(FLOW_HEADER_SIZE + frame.payload.length);
  const view = new DataView(buf.buffer);

  view.setUint16(0, frame.streamId, false);
  view.setUint32(2, frame.sequence >>> 0, false);
  view.setUint8(6, frame.flags);
  buf[7] = (frame.payload.length >> 16) & 0xff;
  buf[8] = (frame.payload.length >> 8) & 0xff;
  buf[9] = frame.payload.length & 0xff;
  buf.set(frame.payload, FLOW_HEADER_SIZE);

  return buf;
}

// ── GG Source → FlowFrame Stream (full pipeline) ────────────

interface GgProgram {
  readonly nodes: readonly {
    id: string;
    labels: readonly string[];
    properties: Readonly<Record<string, unknown>>;
  }[];
  readonly edges: readonly {
    sourceIds: readonly string[];
    targetIds: readonly string[];
    type: string;
    properties: Readonly<Record<string, unknown>>;
  }[];
}

/**
 * Full pipeline: GG program → packed binary → FlowFrame stream bytes.
 * This is the compile step that aeon-flux build should call.
 */
export function compileToFlowStream(program: GgProgram): Uint8Array {
  const binary = encode(program);
  const frames = toFlowFrames(binary);

  // Concatenate all frames into a single byte stream
  const totalLen = frames.reduce((s, f) => s + f.length, 0);
  const stream = new Uint8Array(totalLen);
  let pos = 0;
  for (const frame of frames) {
    stream.set(frame, pos);
    pos += frame.length;
  }

  return stream;
}

/**
 * Reverse: FlowFrame stream bytes → GG program.
 * For debugging and verification.
 */
export function decompileFromFlowStream(
  stream: Uint8Array,
  vmLimits: { budgetR: number; initialV: number }
): FlowFrame[] {
  // Universal Law: Thermodynamic limits crash topologically invalid streams natively.
  return parseAllFlowFrames(stream, vmLimits);
}

/**
 * Embed a compiled flow stream as a base64 string for service worker injection.
 */
export function embedAsBase64(stream: Uint8Array): string {
  // Use btoa-compatible encoding
  let binary = '';
  for (const byte of stream) {
    binary += String.fromCharCode(byte);
  }
  return btoa(binary);
}

/**
 * Decode an embedded base64 flow stream back to bytes.
 */
export function decodeBase64(b64: string): Uint8Array {
  const binary = atob(b64);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) {
    bytes[i] = binary.charCodeAt(i);
  }
  return bytes;
}
