// @a0n/gnosis/src/pattern-horizon-compiler.ts
import {
  CERTIFIED_ALLOCATION_KERNEL_THEOREM_IDS,
  deriveCertifiedAllocationKernel,
  type CertifiedAllocationKernelBundle,
  type CertifiedAllocationKernelInput,
  type CertifiedAllocationKernelTheoremId,
} from './certified-allocation-kernel';
import {
  JUGGLING_CHAOS_ORCHESTRATION_THEOREM_IDS,
  orchestrateJugglingChaosComposition,
  type JugglingChaosOrchestrationBundle,
  type JugglingChaosOrchestrationInput,
  type JugglingChaosOrchestrationTheoremId,
} from './juggling-chaos-orchestration';
import {
  JUGGLING_ORBIT_THEOREM_IDS,
  deriveJugglingOrbitFormalization,
  type JugglingHand,
  type JugglingOrbitFormalizationBundle,
  type JugglingOrbitFormalizationInput,
  type JugglingOrbitNormalizedThrow,
  type JugglingOrbitTheoremId,
  type JugglingOrbitThrowSignal,
} from './juggling-orbit-formalization';
import {
  normalizeFinite,
  normalizePositiveInt,
  positiveModulo,
  quantize,
  nowIso,
} from './juggling-utils';

export const PATTERN_HORIZON_COMPILER_THEOREM_IDS = [
  'THM-PATTERN-HORIZON-SOURCE-DETERMINISTIC',
  'THM-PATTERN-HORIZON-WINDOW-COMPILATION',
  'THM-PATTERN-HORIZON-MACHINE-BOUNDARY-FILTER',
  'THM-PATTERN-HORIZON-OBSERVATION-AUTHORITY',
  'THM-PATTERN-HORIZON-COMMIT-OR-FAIL-CLOSED',
] as const;

export type PatternHorizonCompilerTheoremId =
  (typeof PATTERN_HORIZON_COMPILER_THEOREM_IDS)[number];

export const PATTERN_HORIZON_POLICIES = [
  'hold-safe-state',
  'commit-primary',
  'hedge-shadow',
  'escalate-machine-only',
] as const;

export type PatternHorizonPolicyId = (typeof PATTERN_HORIZON_POLICIES)[number];

export type PatternHorizonSuppressionReason =
  | 'outside-horizon'
  | 'machine-only';

export interface PatternHorizonWindow {
  readonly windowId: string;
  readonly candidateId: number | null;
  readonly throwId: string;
  readonly releaseBeat: number;
  readonly catchBeat: number;
  readonly releaseBeatOffset: number;
  readonly catchBeatOffset: number;
  readonly catchWindowOpenBeatOffset: number;
  readonly catchWindowCloseBeatOffset: number;
  readonly fromHand: JugglingHand;
  readonly toHand: JugglingHand;
  readonly flightBeats: number;
  readonly lane: number;
  readonly angleDeg: number;
  readonly speed: number;
  readonly label: string;
  readonly machineOnlyCandidate: boolean;
  readonly asymmetricMultiplex: boolean;
  readonly admissible: boolean;
  readonly suppressionReason: PatternHorizonSuppressionReason | null;
}

export type PatternHorizonObservationAuthorityInput = Omit<
  CertifiedAllocationKernelInput,
  'generatedAt' | 'source' | 'candidateIds'
>;

interface PatternHorizonBaseInput {
  readonly generatedAt?: string;
  readonly source?: string;
  readonly phaseBeat?: number;
  readonly horizonBeats?: number;
  readonly catchToleranceBeats?: number;
  readonly allowMachineOnly?: boolean;
  readonly observationAuthority: PatternHorizonObservationAuthorityInput;
}

export interface PatternHorizonDirectCompilerInput
  extends PatternHorizonBaseInput {
  readonly kind: 'direct';
  readonly orbit: JugglingOrbitFormalizationInput;
}

export interface PatternHorizonComposedCompilerInput
  extends PatternHorizonBaseInput {
  readonly kind: 'compose';
  readonly composition: JugglingChaosOrchestrationInput;
}

export type PatternHorizonCompilerInput =
  | PatternHorizonDirectCompilerInput
  | PatternHorizonComposedCompilerInput;

export interface PatternHorizonActivePattern {
  readonly kind: 'direct' | 'compose';
  readonly period: number;
  readonly throws: readonly JugglingOrbitThrowSignal[];
  readonly winnerCandidateId: string | null;
  readonly winnerSignature: string | null;
}

export interface PatternHorizonFrontierEntry {
  readonly slot: number;
  readonly candidateId: number;
  readonly window: PatternHorizonWindow;
}

export interface PatternHorizonProgramAggregate {
  readonly phaseBeat: number;
  readonly horizonBeats: number;
  readonly catchToleranceBeats: number;
  readonly totalWindowCount: number;
  readonly admissibleWindowCount: number;
  readonly candidateCount: number;
  readonly suppressedMachineOnlyCount: number;
  readonly suppressedOutsideHorizonCount: number;
  readonly machineOnlyWindowCount: number;
  readonly openTransitionCount: number;
}

export interface PatternHorizonCompilerAggregate
  extends PatternHorizonProgramAggregate {
  readonly observationReady: boolean;
  readonly committed: boolean;
}

export interface PatternHorizonProgram {
  readonly generatedAt: string;
  readonly source: string;
  readonly theoremIds: readonly PatternHorizonCompilerTheoremId[];
  readonly dependentTheoremIds: {
    readonly orbit: readonly JugglingOrbitTheoremId[];
    readonly allocation: readonly CertifiedAllocationKernelTheoremId[];
    readonly composition: readonly JugglingChaosOrchestrationTheoremId[];
  };
  readonly activePattern: PatternHorizonActivePattern;
  readonly orbit: JugglingOrbitFormalizationBundle;
  readonly composition: JugglingChaosOrchestrationBundle | null;
  readonly windows: readonly PatternHorizonWindow[];
  readonly frontier: readonly PatternHorizonFrontierEntry[];
  readonly aggregate: PatternHorizonProgramAggregate;
}

export interface PatternHorizonAdmissionAggregate {
  readonly frontierSlotCount: number;
  readonly observationReady: boolean;
  readonly committed: boolean;
}

export interface PatternHorizonAdmission {
  readonly generatedAt: string;
  readonly source: string;
  readonly theoremIds: readonly PatternHorizonCompilerTheoremId[];
  readonly authority: CertifiedAllocationKernelBundle;
  readonly selectedSlot: number | null;
  readonly selectedWindow: PatternHorizonWindow | null;
  readonly policy: PatternHorizonPolicyId;
  readonly aggregate: PatternHorizonAdmissionAggregate;
}

export interface PatternHorizonCompilerBundle {
  readonly generatedAt: string;
  readonly source: string;
  readonly theoremIds: readonly PatternHorizonCompilerTheoremId[];
  readonly dependentTheoremIds: {
    readonly orbit: readonly JugglingOrbitTheoremId[];
    readonly allocation: readonly CertifiedAllocationKernelTheoremId[];
    readonly composition: readonly JugglingChaosOrchestrationTheoremId[];
  };
  readonly activePattern: PatternHorizonActivePattern;
  readonly orbit: JugglingOrbitFormalizationBundle;
  readonly composition: JugglingChaosOrchestrationBundle | null;
  readonly windows: readonly PatternHorizonWindow[];
  readonly frontier: readonly PatternHorizonFrontierEntry[];
  readonly authority: CertifiedAllocationKernelBundle;
  readonly selectedWindow: PatternHorizonWindow | null;
  readonly policy: PatternHorizonPolicyId;
  readonly aggregate: PatternHorizonCompilerAggregate;
}

interface PatternResolution {
  readonly activePattern: PatternHorizonActivePattern;
  readonly orbit: JugglingOrbitFormalizationBundle;
  readonly composition: JugglingChaosOrchestrationBundle | null;
  readonly compositionTheoremIds: readonly JugglingChaosOrchestrationTheoremId[];
}

interface PendingWindow {
  readonly throwSignal: JugglingOrbitNormalizedThrow;
  readonly releaseBeatOffset: number;
  readonly catchBeat: number;
  readonly catchBeatOffset: number;
  readonly machineOnlyCandidate: boolean;
  readonly asymmetricMultiplex: boolean;
  readonly admissible: boolean;
  readonly suppressionReason: PatternHorizonSuppressionReason | null;
}

const HAND_ORDER: Record<JugglingHand, number> = {
  left: 0,
  right: 1,
};

function compareHands(left: JugglingHand, right: JugglingHand): number {
  return HAND_ORDER[left] - HAND_ORDER[right];
}

function resolvePhaseBeat(value: number | undefined, period: number): number {
  return positiveModulo(Math.floor(normalizeFinite(value, 0)), period);
}

function resolveHorizonBeats(
  value: number | undefined,
  period: number
): number {
  return Math.max(1, Math.min(period, normalizePositiveInt(value, period)));
}

function resolveCatchTolerance(value: number | undefined): number {
  return quantize(Math.max(0, normalizeFinite(value, 0.25)), 3);
}

function collectThrowWitnessSets(orbit: JugglingOrbitFormalizationBundle): {
  readonly machineOnlyThrowIds: ReadonlySet<string>;
  readonly asymmetricThrowIds: ReadonlySet<string>;
} {
  const machineOnlyThrowIds = new Set<string>();
  const asymmetricThrowIds = new Set<string>();

  for (const witness of orbit.asymmetricMultiplexWitnesses) {
    if (witness.asymmetric) {
      for (const throwId of witness.throwIds) {
        asymmetricThrowIds.add(throwId);
      }
    }
    if (witness.machineOnlyCandidate) {
      for (const throwId of witness.throwIds) {
        machineOnlyThrowIds.add(throwId);
      }
    }
  }

  return {
    machineOnlyThrowIds,
    asymmetricThrowIds,
  };
}

function sortPendingWindows(left: PendingWindow, right: PendingWindow): number {
  return (
    left.catchBeatOffset - right.catchBeatOffset ||
    left.releaseBeatOffset - right.releaseBeatOffset ||
    compareHands(left.throwSignal.fromHand, right.throwSignal.fromHand) ||
    compareHands(left.throwSignal.toHand, right.throwSignal.toHand) ||
    left.throwSignal.lane - right.throwSignal.lane ||
    left.throwSignal.id.localeCompare(right.throwSignal.id)
  );
}

/**
 * Handles the gnosis compile Pattern Horizon Windows workflow.
 */
export function compilePatternHorizonWindows(
  orbit: JugglingOrbitFormalizationBundle,
  phaseBeat = 0,
  horizonBeats = orbit.period,
  catchToleranceBeats = 0.25,
  allowMachineOnly = false
): readonly PatternHorizonWindow[] {
  const normalizedPhaseBeat = resolvePhaseBeat(phaseBeat, orbit.period);
  const normalizedHorizonBeats = resolveHorizonBeats(
    horizonBeats,
    orbit.period
  );
  const normalizedCatchTolerance = resolveCatchTolerance(catchToleranceBeats);
  const { machineOnlyThrowIds, asymmetricThrowIds } =
    collectThrowWitnessSets(orbit);

  const pending = orbit.normalizedThrows
    .map((throwSignal) => {
      const catchBeat = positiveModulo(
        throwSignal.beat + throwSignal.flightBeats,
        orbit.period
      );
      const releaseBeatOffset = positiveModulo(
        throwSignal.beat - normalizedPhaseBeat,
        orbit.period
      );
      const catchBeatOffset = positiveModulo(
        catchBeat - normalizedPhaseBeat,
        orbit.period
      );
      const machineOnlyCandidate = machineOnlyThrowIds.has(throwSignal.id);
      const outsideHorizon = catchBeatOffset >= normalizedHorizonBeats;
      const suppressionReason = outsideHorizon
        ? 'outside-horizon'
        : !allowMachineOnly && machineOnlyCandidate
        ? 'machine-only'
        : null;

      return {
        throwSignal,
        releaseBeatOffset,
        catchBeat,
        catchBeatOffset,
        machineOnlyCandidate,
        asymmetricMultiplex: asymmetricThrowIds.has(throwSignal.id),
        admissible: suppressionReason === null,
        suppressionReason,
      } satisfies PendingWindow;
    })
    .sort(sortPendingWindows);

  let nextCandidateId = 1;
  return pending.map((window) => {
    const candidateId = window.admissible ? nextCandidateId++ : null;
    return {
      windowId: `window:${window.throwSignal.id}`,
      candidateId,
      throwId: window.throwSignal.id,
      releaseBeat: window.throwSignal.beat,
      catchBeat: window.catchBeat,
      releaseBeatOffset: window.releaseBeatOffset,
      catchBeatOffset: window.catchBeatOffset,
      catchWindowOpenBeatOffset: quantize(
        Math.max(0, window.catchBeatOffset - normalizedCatchTolerance),
        3
      ),
      catchWindowCloseBeatOffset: quantize(
        window.catchBeatOffset + normalizedCatchTolerance,
        3
      ),
      fromHand: window.throwSignal.fromHand,
      toHand: window.throwSignal.toHand,
      flightBeats: window.throwSignal.flightBeats,
      lane: window.throwSignal.lane,
      angleDeg: window.throwSignal.angleDeg,
      speed: window.throwSignal.speed,
      label: window.throwSignal.label,
      machineOnlyCandidate: window.machineOnlyCandidate,
      asymmetricMultiplex: window.asymmetricMultiplex,
      admissible: window.admissible,
      suppressionReason: window.suppressionReason,
    } satisfies PatternHorizonWindow;
  });
}

function resolvePattern(
  input: PatternHorizonCompilerInput,
  generatedAt: string,
  source: string
): PatternResolution {
  if (input.kind === 'direct') {
    const orbit = deriveJugglingOrbitFormalization({
      ...input.orbit,
      generatedAt: input.orbit.generatedAt ?? generatedAt,
      source: input.orbit.source ?? `${source}:orbit`,
    });
    return {
      activePattern: {
        kind: 'direct',
        period: orbit.period,
        throws: input.orbit.throws,
        winnerCandidateId: null,
        winnerSignature: null,
      },
      orbit,
      composition: null,
      compositionTheoremIds: [],
    };
  }

  const composition = orchestrateJugglingChaosComposition({
    ...input.composition,
    generatedAt: input.composition.generatedAt ?? generatedAt,
    source: input.composition.source ?? `${source}:composition`,
  });

  return {
    activePattern: {
      kind: 'compose',
      period: composition.period,
      throws: composition.winner.throws,
      winnerCandidateId: composition.winner.candidateId,
      winnerSignature: composition.winner.signature,
    },
    orbit: composition.winner.orbit,
    composition,
    compositionTheoremIds: JUGGLING_CHAOS_ORCHESTRATION_THEOREM_IDS,
  };
}

function resolveSelectedWindow(
  windows: readonly PatternHorizonWindow[],
  authority: CertifiedAllocationKernelBundle
): PatternHorizonWindow | null {
  if (authority.decision.kind !== 'allocate') {
    return null;
  }
  const candidateId = authority.decision.candidateId;

  return windows.find((window) => window.candidateId === candidateId) ?? null;
}

function resolveSelectedFrontierEntry(
  frontier: readonly PatternHorizonFrontierEntry[],
  authority: CertifiedAllocationKernelBundle
): PatternHorizonFrontierEntry | null {
  if (authority.decision.kind !== 'allocate') {
    return null;
  }
  const candidateId = authority.decision.candidateId;

  return frontier.find((entry) => entry.slot === candidateId) ?? null;
}

function resolvePolicy(
  selectedWindow: PatternHorizonWindow | null,
  authority: CertifiedAllocationKernelBundle,
  orbit: JugglingOrbitFormalizationBundle,
  composition: JugglingChaosOrchestrationBundle | null
): PatternHorizonPolicyId {
  if (authority.decision.kind !== 'allocate' || selectedWindow === null) {
    return 'hold-safe-state';
  }

  if (selectedWindow.machineOnlyCandidate) {
    return 'escalate-machine-only';
  }

  const uncertaintyLoad = composition?.winner.score.mpc.uncertaintyLoad ?? 0;
  const nonConvexRisk = composition?.winner.score.mpc.nonConvexRisk ?? 0;
  if (
    orbit.aggregate.openTransitionCount > 0 ||
    uncertaintyLoad >= 0.5 ||
    nonConvexRisk >= 0.5
  ) {
    return 'hedge-shadow';
  }

  return 'commit-primary';
}

function buildProgramAggregate(
  windows: readonly PatternHorizonWindow[],
  phaseBeat: number,
  horizonBeats: number,
  catchToleranceBeats: number,
  orbit: JugglingOrbitFormalizationBundle
): PatternHorizonProgramAggregate {
  return {
    phaseBeat,
    horizonBeats,
    catchToleranceBeats,
    totalWindowCount: windows.length,
    admissibleWindowCount: windows.filter((window) => window.admissible).length,
    candidateCount: windows.flatMap((window) =>
      window.candidateId === null ? [] : [window.candidateId]
    ).length,
    suppressedMachineOnlyCount: windows.filter(
      (window) => window.suppressionReason === 'machine-only'
    ).length,
    suppressedOutsideHorizonCount: windows.filter(
      (window) => window.suppressionReason === 'outside-horizon'
    ).length,
    machineOnlyWindowCount: windows.filter(
      (window) => window.machineOnlyCandidate
    ).length,
    openTransitionCount: orbit.aggregate.openTransitionCount,
  };
}

/**
 * Handles the gnosis compile Pattern Horizon Program workflow.
 */
export function compilePatternHorizonProgram(
  input: PatternHorizonCompilerInput
): PatternHorizonProgram {
  const generatedAt = input.generatedAt ?? nowIso();
  const source = input.source?.trim() || 'pattern-horizon-compiler';
  const resolution = resolvePattern(input, generatedAt, source);
  const phaseBeat = resolvePhaseBeat(input.phaseBeat, resolution.orbit.period);
  const horizonBeats = resolveHorizonBeats(
    input.horizonBeats,
    resolution.orbit.period
  );
  const catchToleranceBeats = resolveCatchTolerance(input.catchToleranceBeats);
  const windows = compilePatternHorizonWindows(
    resolution.orbit,
    phaseBeat,
    horizonBeats,
    catchToleranceBeats,
    input.allowMachineOnly ?? false
  );
  const frontier = windows.flatMap((window) =>
    window.candidateId === null
      ? []
      : [
          {
            slot: windows
              .filter((candidate) => candidate.candidateId !== null)
              .findIndex((candidate) => candidate.windowId === window.windowId),
            candidateId: window.candidateId,
            window,
          } satisfies PatternHorizonFrontierEntry,
        ]
  );

  return {
    generatedAt,
    source,
    theoremIds: PATTERN_HORIZON_COMPILER_THEOREM_IDS,
    dependentTheoremIds: {
      orbit: JUGGLING_ORBIT_THEOREM_IDS,
      allocation: CERTIFIED_ALLOCATION_KERNEL_THEOREM_IDS,
      composition: resolution.compositionTheoremIds,
    },
    activePattern: resolution.activePattern,
    orbit: resolution.orbit,
    composition: resolution.composition,
    windows,
    frontier,
    aggregate: buildProgramAggregate(
      windows,
      phaseBeat,
      horizonBeats,
      catchToleranceBeats,
      resolution.orbit
    ),
  };
}

/**
 * Handles the gnosis admit Pattern Horizon Program workflow.
 */
export function admitPatternHorizonProgram(
  program: PatternHorizonProgram,
  observationAuthority: PatternHorizonObservationAuthorityInput
): PatternHorizonAdmission {
  const authority = deriveCertifiedAllocationKernel({
    ...observationAuthority,
    generatedAt: program.generatedAt,
    source: program.source,
    candidateIds: program.frontier.map((entry) => entry.slot),
  });
  const selectedEntry = resolveSelectedFrontierEntry(
    program.frontier,
    authority
  );
  const selectedSlot = selectedEntry?.slot ?? null;
  const selectedWindow = selectedEntry?.window ?? null;
  const policy = resolvePolicy(
    selectedWindow,
    authority,
    program.orbit,
    program.composition
  );

  return {
    generatedAt: program.generatedAt,
    source: program.source,
    theoremIds: PATTERN_HORIZON_COMPILER_THEOREM_IDS,
    authority,
    selectedSlot,
    selectedWindow,
    policy,
    aggregate: {
      frontierSlotCount: program.frontier.length,
      observationReady: authority.aggregate.ready,
      committed: authority.decision.kind === 'allocate',
    },
  };
}

/**
 * Handles the gnosis derive Pattern Horizon Compiler workflow.
 */
export function derivePatternHorizonCompiler(
  input: PatternHorizonCompilerInput
): PatternHorizonCompilerBundle {
  const program = compilePatternHorizonProgram(input);
  const admission = admitPatternHorizonProgram(
    program,
    input.observationAuthority
  );
  const authority = deriveCertifiedAllocationKernel({
    ...input.observationAuthority,
    generatedAt: program.generatedAt,
    source: program.source,
    candidateIds: program.windows.flatMap((window) =>
      window.candidateId === null ? [] : [window.candidateId]
    ),
  });
  const selectedWindow =
    admission.selectedWindow ??
    resolveSelectedWindow(program.windows, authority);

  return {
    generatedAt: program.generatedAt,
    source: program.source,
    theoremIds: program.theoremIds,
    dependentTheoremIds: program.dependentTheoremIds,
    activePattern: program.activePattern,
    orbit: program.orbit,
    composition: program.composition,
    windows: program.windows,
    frontier: program.frontier,
    authority,
    selectedWindow,
    policy: resolvePolicy(
      selectedWindow,
      authority,
      program.orbit,
      program.composition
    ),
    aggregate: {
      ...program.aggregate,
      observationReady: admission.aggregate.observationReady,
      committed: authority.decision.kind === 'allocate',
    },
  };
}
