// @a0n/gnosis/src/phyle-temporal-depth.ts
/**
 * phyle-temporal-depth.ts — Noise-resolved depth reconstruction via triton windows.
 *
 * Orchestrates the Phyle temporal accumulation pipeline:
 * - Groups frames into triton windows (past/present/future)
 * - Drives PhyleCell fork/race/fold scheduling per window
 * - Classifies noise color and promotes standing waves
 * - Integrates deblur stability across time
 *
 * The key insight: information IS the difference relation across three time
 * samples. Single-shot depth is structurally incomplete. We rotate space
 * through time, and constructive interference builds resolution naturally.
 */

import {
  type PhyleCell,
  type PhyleMesh,
  PHYLE_DIRECTIONS,
  CLOCK_PERIOD,
  createPhyleCell,
  createPhyleMesh,
} from './phyle-cell.js';

// ─── Types ──────────────────────────────────────────────────────────

export type TritonPhase = 'past' | 'present' | 'future';

export const TRITON_PHASES: readonly TritonPhase[] = ['past', 'present', 'future'] as const;
export const PERFECT_FIFTH_STRIDE = 7;

export interface DepthFrame {
  values: Float32Array;
  confidence: Float32Array;
  deblurAdmission: Float32Array;
  width: number;
  height: number;
  frameIndex: number;
  timestamp: number;
}

export interface TritonWindow {
  frames: [DepthFrame | null, DepthFrame | null, DepthFrame | null];
  filled: [boolean, boolean, boolean];
  rotationDirection: number;
  windowIndex: number;
}

export type NoiseColor = 'white' | 'pink' | 'brown' | 'standingWave';

export interface TritonWindowResult {
  windowIndex: number;
  shell: number;
  standingWaveCount: number;
  pinkResolved: number;
  brownResolved: number;
  whiteDeferred: number;
  rotationDirection: number;
  meanSecondDegreeDiff: number;
  meanDeblurStability: number;
}

export interface PhyleTemporalPipelineState {
  activeWindow: TritonWindow;
  completedWindows: number;
  cell: PhyleCell;
  currentShell: number;
  fusedDepth: Float32Array | null;
  standingWaveMask: Float32Array | null;
  deblurStability: Float32Array | null;
  width: number;
  height: number;
}

// ─── Noise color spectral budget ────────────────────────────────────

/**
 * Handles the gnosis noise Alpha workflow.
 */
export function noiseAlpha(color: NoiseColor): number {
  switch (color) {
    case 'white': return 0;
    case 'pink': return 1;
    case 'brown': return 2;
    case 'standingWave': return 0;
  }
}

/**
 * Handles the gnosis spectral Budget workflow.
 */
export function spectralBudget(shell: number): number {
  return shell + 1;
}

/**
 * Handles the gnosis noise Resolves At Shell workflow.
 */
export function noiseResolvesAtShell(color: NoiseColor): number | null {
  switch (color) {
    case 'white': return null;
    case 'pink': return 0;
    case 'brown': return 1;
    case 'standingWave': return 0;
  }
}

/**
 * Handles the gnosis noise Admitted At Shell workflow.
 */
export function noiseAdmittedAtShell(color: NoiseColor, shell: number): boolean {
  return noiseAlpha(color) <= spectralBudget(shell);
}

/**
 * Handles the gnosis frames At Shell workflow.
 */
export function framesAtShell(shell: number): number {
  return CLOCK_PERIOD * Math.pow(CLOCK_PERIOD, shell);
}

/**
 * Handles the gnosis rotation Direction workflow.
 */
export function rotationDirection(windowIndex: number): number {
  return (windowIndex * PERFECT_FIFTH_STRIDE) % PHYLE_DIRECTIONS;
}

// ─── Pipeline ───────────────────────────────────────────────────────

/**
 * Creates the Phyle Temporal Pipeline.
 */
export function createPhyleTemporalPipeline(width: number, height: number): PhyleTemporalPipelineState {
  return {
    activeWindow: createTritonWindow(0, 0),
    completedWindows: 0,
    cell: createPhyleCell(),
    currentShell: 0,
    fusedDepth: null,
    standingWaveMask: null,
    deblurStability: null,
    width,
    height,
  };
}

function createTritonWindow(windowIndex: number, completedWindows: number): TritonWindow {
  return {
    frames: [null, null, null],
    filled: [false, false, false],
    rotationDirection: rotationDirection(completedWindows),
    windowIndex,
  };
}

/**
 * Handles the gnosis ingest Frame workflow.
 */
export function ingestFrame(
  state: PhyleTemporalPipelineState,
  frame: DepthFrame,
): TritonWindowResult | null {
  if (frame.width !== state.width || frame.height !== state.height) {
    throw new Error(
      `Frame dimensions ${frame.width}x${frame.height} do not match pipeline ${state.width}x${state.height}`
    );
  }

  const phaseIndex = state.activeWindow.filled.indexOf(false);
  if (phaseIndex === -1) {
    throw new Error('Triton window already complete — call completeWindow first');
  }

  state.activeWindow.frames[phaseIndex] = frame;
  state.activeWindow.filled[phaseIndex] = true;

  if (state.activeWindow.filled[0] && state.activeWindow.filled[1] && state.activeWindow.filled[2]) {
    return completeTritonWindow(state);
  }

  return null;
}

function completeTritonWindow(state: PhyleTemporalPipelineState): TritonWindowResult {
  const pix = state.width * state.height;
  const [past, present, future] = state.activeWindow.frames as [DepthFrame, DepthFrame, DepthFrame];

  if (!state.fusedDepth) state.fusedDepth = new Float32Array(pix);
  if (!state.standingWaveMask) state.standingWaveMask = new Float32Array(pix);
  if (!state.deblurStability) state.deblurStability = new Float32Array(pix);

  const standingWaveThreshold = 0.001;
  const deblurThreshold = 0.01;

  let standingWaveCount = 0;
  let pinkResolved = 0;
  let brownResolved = 0;
  let whiteDeferred = 0;
  let sumSecondDegree = 0;
  let sumDeblurStability = 0;

  const shell = state.currentShell;

  for (let i = 0; i < pix; i++) {
    const obs: [number, number, number] = [past.values[i]!, present.values[i]!, future.values[i]!];
    const conf: [number, number, number] = [past.confidence[i]!, present.confidence[i]!, future.confidence[i]!];
    const deblur: [number, number, number] = [past.deblurAdmission[i]!, present.deblurAdmission[i]!, future.deblurAdmission[i]!];

    // First degree diffs
    const dPP = Math.abs(obs[0] - obs[1]);
    const dPF = Math.abs(obs[1] - obs[2]);
    const secondDegreeDiff = Math.abs(dPP - dPF);

    // Phase variance
    const mean = (obs[0] + obs[1] + obs[2]) / 3;
    const phaseVariance = ((obs[0] - mean) ** 2 + (obs[1] - mean) ** 2 + (obs[2] - mean) ** 2) / 3;

    // Deblur variance
    const dMean = (deblur[0] + deblur[1] + deblur[2]) / 3;
    const deblurVariance = ((deblur[0] - dMean) ** 2 + (deblur[1] - dMean) ** 2 + (deblur[2] - dMean) ** 2) / 3;

    sumSecondDegree += secondDegreeDiff;
    sumDeblurStability += 1 - Math.min(deblurVariance, 1);

    // Classify noise color
    let color: NoiseColor;
    if (phaseVariance < standingWaveThreshold && deblurVariance < deblurThreshold) {
      color = 'standingWave';
    } else if (phaseVariance < standingWaveThreshold * 10) {
      color = 'pink';
    } else if (phaseVariance < standingWaveThreshold * 100) {
      color = 'brown';
    } else {
      color = 'white';
    }

    // Try resolve
    const requiredShell = noiseResolvesAtShell(color);
    if (requiredShell !== null && shell >= requiredShell) {
      const totalConf = conf[0] + conf[1] + conf[2];
      if (totalConf > 0) {
        state.fusedDepth[i] = (obs[0] * conf[0] + obs[1] * conf[1] + obs[2] * conf[2]) / totalConf;
      } else {
        state.fusedDepth[i] = mean;
      }
      state.deblurStability[i] = 1 - Math.min(deblurVariance, 1);

      switch (color) {
        case 'standingWave':
          state.standingWaveMask[i] = 1.0;
          standingWaveCount++;
          break;
        case 'pink':
          state.standingWaveMask[i] = 0.8;
          pinkResolved++;
          break;
        case 'brown':
          state.standingWaveMask[i] = 0.6;
          brownResolved++;
          break;
      }
    } else if (color === 'white') {
      whiteDeferred++;
    }
  }

  // Advance state
  state.completedWindows++;

  // Shell promotion
  const windowsForNextShell = Math.pow(CLOCK_PERIOD, state.currentShell + 1);
  if (state.completedWindows >= windowsForNextShell && state.currentShell < 3) {
    state.currentShell++;
  }

  // Reset active window
  state.activeWindow = createTritonWindow(state.completedWindows, state.completedWindows);

  return {
    windowIndex: state.completedWindows - 1,
    shell,
    standingWaveCount,
    pinkResolved,
    brownResolved,
    whiteDeferred,
    rotationDirection: state.activeWindow.rotationDirection,
    meanSecondDegreeDiff: sumSecondDegree / pix,
    meanDeblurStability: sumDeblurStability / pix,
  };
}

// ─── Temporal deblur classification ─────────────────────────────────

export interface TemporalDeblurReport {
  stableEdgeCount: number;
  noiseArtifactCount: number;
  stableEdgeMask: Float32Array;
  edgeConfidence: Float32Array;
}

/**
 * Handles the gnosis classify Temporal Deblur workflow.
 */
export function classifyTemporalDeblur(state: PhyleTemporalPipelineState): TemporalDeblurReport | null {
  if (!state.deblurStability || !state.standingWaveMask) return null;

  const pix = state.width * state.height;
  const stableEdgeMask = new Float32Array(pix);
  const edgeConfidence = new Float32Array(pix);
  let stableEdgeCount = 0;
  let noiseArtifactCount = 0;

  for (let i = 0; i < pix; i++) {
    const stability = state.deblurStability[i]!;
    const swMask = state.standingWaveMask[i]!;

    if (stability > 0.9 && swMask > 0.5) {
      stableEdgeMask[i] = stability;
      edgeConfidence[i] = stability * swMask;
      stableEdgeCount++;
    } else if (stability < 0.5) {
      noiseArtifactCount++;
    }
  }

  return { stableEdgeCount, noiseArtifactCount, stableEdgeMask, edgeConfidence };
}

// ─── Pipeline reporting ─────────────────────────────────────────────

export interface PipelineReport {
  framesIngested: number;
  tritonWindowsCompleted: number;
  currentShell: number;
  standingWaveCoverage: number;
  framesUntilNextShell: number;
  noiseColorDistribution: Record<NoiseColor, number>;
}

/**
 * Handles the gnosis pipeline Report workflow.
 */
export function pipelineReport(state: PhyleTemporalPipelineState): PipelineReport {
  const pix = state.width * state.height;
  const framesIngested = state.completedWindows * CLOCK_PERIOD +
    state.activeWindow.filled.filter(Boolean).length;

  let standingWavePixels = 0;
  if (state.standingWaveMask) {
    for (let i = 0; i < pix; i++) {
      if (state.standingWaveMask[i]! > 0) standingWavePixels++;
    }
  }

  const windowsForNextShell = Math.pow(CLOCK_PERIOD, state.currentShell + 1);
  const framesUntilNextShell = state.currentShell < 3
    ? (windowsForNextShell - state.completedWindows) * CLOCK_PERIOD
    : 0;

  return {
    framesIngested,
    tritonWindowsCompleted: state.completedWindows,
    currentShell: state.currentShell,
    standingWaveCoverage: pix > 0 ? standingWavePixels / pix : 0,
    framesUntilNextShell,
    noiseColorDistribution: { white: 0, pink: 0, brown: 0, standingWave: 0 },
  };
}
