// @a0n/gnosis/src/phyle-cell.ts
/**
 * Phyle Cell -- The Foundational Scheduling Primitive
 *
 * A Phyle is 9 vertices arranged as a tripod-of-tripods (3 radials x 3 rings).
 * The hub is the fold collapse point, separate from the 9 carrier bars.
 * 3 bond endpoints connect cells to each other in the PhyleMesh.
 *
 * All operations are O(1) via fixed-size array access -- no adjacency traversal.
 * The 27-direction race uses modular rotation over 3^3 orientations.
 *
 * Mirrors the Lean proofs in gnosis-math/Gnosis/Phyle.lean and
 * PhyleLightEmission.lean.
 *
 * @packageDocumentation
 */

import { godFormula } from './holy-trinity';
import { type Trit, FORK, VOID, VENT } from './scheduling-mesh';

// ─── Constants (matching Lean proofs) ─────────────────────────────────

/** Number of carrier bars in a Phyle cell. */
export const PHYLE_BARS = 9;

/** Bars in the old triangle (one tripod arm). */
export const OLD_TRIANGLE_BARS = 3;

/** Margin bars beyond the old triangle: 9 - 3 = 6. */
export const PHYLE_MARGIN = 6;

/** Light ordinal: each tripod arm has 3 vertices. */
export const LIGHT_ORDINAL = 3;

/** Total race directions: 3^3 = 27. */
export const PHYLE_DIRECTIONS = 27;

/** Triton temporal period: past, present, future. */
export const CLOCK_PERIOD = 3;

/** Emanation shells for LOD rendering. */
export const EMANATION_SHELLS = 3;

/** Total rays: 3^9 = 19683. */
export const PHYLE_RAYS = 19683;

// ─── Tripod Arm Layout ───────────────────────────────────────────────

/** Arm A: positions 0, 1, 2. */
export const ARM_A: readonly [0, 1, 2] = [0, 1, 2] as const;

/** Arm B: positions 3, 4, 5. */
export const ARM_B: readonly [3, 4, 5] = [3, 4, 5] as const;

/** Arm C: positions 6, 7, 8. */
export const ARM_C: readonly [6, 7, 8] = [6, 7, 8] as const;

/** All arms indexed by radial. */
export const TRIPOD_ARMS = [ARM_A, ARM_B, ARM_C] as const;

// ─── Types ───────────────────────────────────────────────────────────

/** A position index within the Phyle cell (0-8). */
export type PhylePosition = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8;

/** A vertex in the Phyle cell. */
export interface PhyleVertex {
  /** Unique vertex ID (globally unique within the mesh). */
  readonly id: number;
  /** Position within the cell (0-8). */
  readonly position: PhylePosition;
  /** God formula weight: w = R - min(v, R) + 1. */
  weight: number;
  /** Rejection count for this vertex. */
  rejections: number;
  /** Current payload (task being processed). */
  payload: unknown | null;
  /** Vertex state as a trit. */
  state: Trit;
  /** Whether this vertex is currently busy. */
  busy: boolean;
}

/** A bond endpoint connecting this cell to another. */
export interface PhyleBond {
  /** Local vertex position this bond attaches to (tip of each arm). */
  readonly localPosition: PhylePosition;
  /** Target cell index in the mesh (-1 if unbound). */
  targetCell: number;
  /** Target position in the remote cell. */
  targetPosition: PhylePosition;
}

// ─── PhyleCell ───────────────────────────────────────────────────────

let nextVertexId = 0;

/** Reset the global vertex ID counter (for testing). */
export function resetVertexIds(): void {
  nextVertexId = 0;
}

/**
 * PhyleCell -- 9 vertices in a tripod-of-tripods, plus a hub.
 *
 * Fixed-size array for O(1) access. No adjacency traversal needed.
 */
export class PhyleCell {
  /** The 9 carrier bar vertices (fixed-size). */
  readonly vertices: readonly [
    PhyleVertex, PhyleVertex, PhyleVertex,
    PhyleVertex, PhyleVertex, PhyleVertex,
    PhyleVertex, PhyleVertex, PhyleVertex,
  ];

  /** Hub vertex -- fold collapse point, separate from the 9 bars. */
  readonly hub: PhyleVertex;

  /** 3 bond endpoints for inter-cell connection (one per arm tip). */
  readonly bonds: readonly [PhyleBond, PhyleBond, PhyleBond];

  /** Rejection round counter for god formula. */
  private rounds: number = 0;

  constructor() {
    const verts: PhyleVertex[] = [];
    for (let i = 0; i < PHYLE_BARS; i++) {
      verts.push({
        id: nextVertexId++,
        position: i as PhylePosition,
        weight: 1,
        rejections: 0,
        payload: null,
        state: VOID,
        busy: false,
      });
    }
    this.vertices = verts as unknown as PhyleCell['vertices'];

    this.hub = {
      id: nextVertexId++,
      position: 0 as PhylePosition,
      weight: 1,
      rejections: 0,
      payload: null,
      state: VOID,
      busy: false,
    };

    // Bond endpoints at the tip of each arm (positions 2, 5, 8).
    this.bonds = [
      { localPosition: 2 as PhylePosition, targetCell: -1, targetPosition: 0 as PhylePosition },
      { localPosition: 5 as PhylePosition, targetCell: -1, targetPosition: 3 as PhylePosition },
      { localPosition: 8 as PhylePosition, targetCell: -1, targetPosition: 6 as PhylePosition },
    ] as const;
  }

  /** Direct O(1) vertex access by position. */
  vertex(pos: PhylePosition): PhyleVertex {
    return this.vertices[pos];
  }

  /**
   * P1 FORK: Write a task to all 9 positions. O(1) -- fixed iteration.
   */
  fork(task: unknown): void {
    this.rounds++;
    for (let i = 0; i < PHYLE_BARS; i++) {
      const v = this.vertices[i];
      (v as { payload: unknown }).payload = task;
      (v as { state: Trit }).state = FORK;
      (v as { busy: boolean }).busy = true;
    }
    (this.hub as { state: Trit }).state = FORK;
  }

  /**
   * P3 RACE: Select the winner by god formula weight. O(1) -- fixed iteration.
   *
   * All vertices evaluate. The highest-weight vertex wins.
   * Losers increment rejections and reweight.
   */
  race<T>(evaluate: (vertex: PhyleVertex) => T): { winner: PhyleVertex; result: T } {
    let winnerId = 0;
    let bestWeight = this.vertices[0].weight;

    // Find the highest-weight vertex.
    for (let i = 1; i < PHYLE_BARS; i++) {
      const w = this.vertices[i].weight;
      if (w > bestWeight) {
        bestWeight = w;
        winnerId = i;
      }
    }

    const winner = this.vertices[winnerId];
    const result = evaluate(winner);

    // Reject losers.
    for (let i = 0; i < PHYLE_BARS; i++) {
      const v = this.vertices[i];
      if (i === winnerId) {
        (v as { state: Trit }).state = VOID;
        (v as { busy: boolean }).busy = false;
      } else {
        (v as { rejections: number }).rejections++;
        (v as { weight: number }).weight = godFormula(this.rounds, v.rejections);
        (v as { state: Trit }).state = VENT;
        (v as { busy: boolean }).busy = false;
        (v as { payload: unknown }).payload = null;
      }
    }

    return { winner, result };
  }

  /**
   * P2 FOLD: Collect from all 9 positions. O(1) -- fixed iteration.
   *
   * Extracts a result from each vertex and resets it to idle.
   */
  fold<T>(extract: (vertex: PhyleVertex) => T): T[] {
    const results: T[] = [];
    for (let i = 0; i < PHYLE_BARS; i++) {
      const v = this.vertices[i];
      if (v.payload !== null) {
        results.push(extract(v));
      }
      (v as { state: Trit }).state = VOID;
      (v as { payload: unknown }).payload = null;
      (v as { busy: boolean }).busy = false;
    }
    (this.hub as { state: Trit }).state = VOID;
    return results;
  }

  /**
   * 27-direction race with modular rotation. O(1) -- 27 fixed steps.
   *
   * Each direction is a (radial, ring, phase-offset) triple from Z/3 x Z/3 x Z/3.
   * The phase parameter rotates the starting orientation.
   */
  raceDirections<T>(
    phase: number,
    evaluate: (vertex: PhyleVertex, direction: number) => T,
  ): { winner: PhyleVertex; result: T; direction: number } {
    let bestWeight = -Infinity;
    let winnerId = 0;
    let winDirection = 0;

    for (let d = 0; d < PHYLE_DIRECTIONS; d++) {
      // Map direction to a vertex via modular rotation.
      const rotated = (d + phase) % PHYLE_DIRECTIONS;
      const pos = (rotated % PHYLE_BARS) as PhylePosition;
      const v = this.vertices[pos];
      if (v.weight > bestWeight) {
        bestWeight = v.weight;
        winnerId = pos;
        winDirection = d;
      }
    }

    const winner = this.vertices[winnerId];
    const result = evaluate(winner, winDirection);
    return { winner, result, direction: winDirection };
  }

  /** Reset all vertices and hub to idle. */
  reset(): void {
    for (let i = 0; i < PHYLE_BARS; i++) {
      const v = this.vertices[i];
      (v as { weight: number }).weight = 1;
      (v as { rejections: number }).rejections = 0;
      (v as { payload: unknown }).payload = null;
      (v as { state: Trit }).state = VOID;
      (v as { busy: boolean }).busy = false;
    }
    (this.hub as { weight: number }).weight = 1;
    (this.hub as { rejections: number }).rejections = 0;
    (this.hub as { payload: unknown }).payload = null;
    (this.hub as { state: Trit }).state = VOID;
    (this.hub as { busy: boolean }).busy = false;
    this.rounds = 0;
  }
}

// ─── PhyleMesh ───────────────────────────────────────────────────────

/**
 * PhyleMesh -- a composition of Phyle cells.
 *
 * Composite dimension = (cellCount - 1) * 9 + 1.
 * Cells are bonded at arm tips for inter-cell propagation.
 */
export class PhyleMesh {
  /** All cells in the mesh. */
  readonly cells: PhyleCell[];

  constructor() {
    this.cells = [];
  }

  /** Number of cells in the mesh. */
  get cellCount(): number {
    return this.cells.length;
  }

  /** Composite dimension: (cellCount - 1) * 9 + 1. */
  get compositeDimension(): number {
    if (this.cells.length <= 0) return 0;
    return (this.cells.length - 1) * PHYLE_BARS + 1;
  }

  /** Create a new cell and add it to the mesh. Bonds to the previous cell. */
  createCell(): PhyleCell {
    const cell = new PhyleCell();
    const idx = this.cells.length;
    this.cells.push(cell);

    // Bond to the previous cell if one exists.
    if (idx > 0) {
      const prev = this.cells[idx - 1]!;
      // Bond arm C tip of prev -> arm A base of new.
      const prevBond = prev.bonds[2] as { targetCell: number; targetPosition: PhylePosition };
      prevBond.targetCell = idx;
      prevBond.targetPosition = 0 as PhylePosition;

      const newBond = cell.bonds[0] as { targetCell: number; targetPosition: PhylePosition };
      newBond.targetCell = idx - 1;
      newBond.targetPosition = 8 as PhylePosition;
    }

    return cell;
  }

  /** Get a cell by index. */
  getCell(index: number): PhyleCell | undefined {
    return this.cells[index];
  }
}

// ─── Named Compositions ──────────────────────────────────────────────

/**
 * Named Phyle compositions from the gnostic hierarchy.
 *
 * sophia  = 1 cell  (dim 1, the monad)
 * kenoma  = 2 cells (dim 10, the dyad)
 * void    = 3 cells (dim 19, the triad)
 * pleroma = 7 cells (dim 55, the heptad -- Fano plane)
 */
export const PHYLE_COMPOSITIONS: ReadonlyMap<string, number> = new Map([
  ['sophia', 1],
  ['kenoma', 2],
  ['void', 3],
  ['pleroma', 7],
]);

// ─── Factory ─────────────────────────────────────────────────────────

/**
 * Create a single Phyle cell.
 *
 * @returns A PhyleCell with fresh vertex identifiers
 */
export function createPhyleCell(): PhyleCell {
  return new PhyleCell();
}

/**
 * Create a PhyleMesh with a given number of cells.
 *
 * @param cells - Number of Phyle cells to compose
 * @returns A bonded PhyleMesh
 */
export function createPhyleMesh(cells: number): PhyleMesh {
  const mesh = new PhyleMesh();
  for (let i = 0; i < cells; i++) {
    mesh.createCell();
  }
  return mesh;
}
