// @a0n/gnosis/src/physical-emotion.ts
/**
 * physical-emotion.ts -- The Thermodynamic Heat of General Emotion
 *
 * Emotion is defined as the rate of change of the complement distribution
 * (the void boundary accumulating or releasing over time).
 * 
 * Formalized in `PhysicalEmotion.lean` (Pass 53), any shift in this boundary
 * constitutes computational labor. Thus, feeling an emotion strictly 
 * requires the dissipation of physical heat and measurable topological waste.
 *
 * Mechanized in PhysicalEmotion.lean (zero sorry):
 *   - emotion_is_heat_dissipation: Emotional magnitude = Landauer Heat.
 *   - emotional_friction_is_waste: Internal conflict = topological waste.
 *   - intense_emotion_is_high_wattage: Larger void shift = higher heat.
 *   - calm_minimizes_entropy_production: Zero shift = irreducible baseline.
 *   - physical_emotion_master_theorem: Master thermodynamic identity.
 *
 * @module @a0n/gnosis/physical-emotion
 */

// ============================================================================
// Core Types
// ============================================================================

/** The void boundary state defining the emotion. */
export interface EmotionalState {
  /** The magnitude of the void boundary at the prior tick (t-1). */
  readonly previousVoid: number;
  /** The magnitude of the void boundary at the current tick (t). */
  readonly currentVoid: number;
}

/** 
 * A generalized Emotional Output bridging subjective feeling to physical physics.
 */
export interface PhysicalEmotionMetrics {
  /** The absolute magnitude of the emotion (the amount of void shifted). */
  readonly emotionalMagnitude: number;
  /** The raw void delta. Positive = accumulating, Negative = releasing. */
  readonly voidAccumulation: number;
  /** 
   * The structural deficit equivalent (Bules of value destroyed/restructured 
   * to accommodate the emotional response). 
   */
  readonly deficitEquivalent: number;
  /** 
   * The computational waste generated navigating the feeling.
   * Emotion feels difficult because it generates systemic friction (waste).
   */
  readonly wasteCycles: number;
  /**
   * The actual Landauer Heat required to execute the emotion in joules or 
   * wattage (scaled to application time). 
   */
  readonly heatQuanta: number;
  /** Lean theorem anchor for the physical generation. */
  readonly leanAnchor: string;
}

// ============================================================================
// Factory Functions
// ============================================================================

/**
 * Computes the thermodynamic load of a given emotional state.
 *
 * Per `emotion_is_heat_dissipation` and `emotional_friction_is_waste`,
 * changing the void boundary inherently costs energy.
 * 
 * @param state The current vs prior void boundaries of the emotion.
 * @param energyConstants Multipliers for mapping Bules to runtime joules/watts.
 */
export function computePhysicalEmotion(
  state: EmotionalState,
  energyConstants: { 
    buleToJoules: number; 
    buleToWaste: number; 
  } = { buleToJoules: 1, buleToWaste: 1 }
): PhysicalEmotionMetrics {
  
  const voidAccumulation = state.currentVoid - state.previousVoid;
  const emotionalMagnitude = Math.abs(voidAccumulation);
  
  // Per buleAsDeficit(magnitude, 1 stream) in Lean
  const deficitEquivalent = emotionalMagnitude; 

  const isCalm = emotionalMagnitude === 0;

  return {
    emotionalMagnitude,
    voidAccumulation,
    deficitEquivalent,
    wasteCycles: deficitEquivalent * energyConstants.buleToWaste,
    heatQuanta: deficitEquivalent * energyConstants.buleToJoules,
    leanAnchor: isCalm ? 'calm_minimizes_entropy_production' : 'emotion_is_heat_dissipation',
  };
}

/**
 * Computes a calm state where zero excess entropy is produced.
 * 
 * Per `calm_minimizes_entropy_production`, maintaining a stable boundary
 * minimizes heat to the baseline.
 */
export function getZenState(currentVoid: number): PhysicalEmotionMetrics {
  return computePhysicalEmotion({ previousVoid: currentVoid, currentVoid });
}
