// @a0n/gnosis/src/physics-sandbox.ts
/**
 * **Not verified by Lean.** Deterministic numeric toys for demos only.
 *
 * For formal content see `lean/Lean/PhysicsSandbox.lean` and
 * `docs/PHYSICS_SANDBOX.md`.
 *
 * @module @a0n/gnosis/physics-sandbox
 */

/** Stable Lean library paths for tooling, logs, and cross-repo trace anchors. */
export const GNOSIS_PHYSICS_SANDBOX_LEAN_MODULES = [
  'PhysicsSandbox',
  'ForkRaceFoldTheorems.ClassicalSandbox',
  'ForkRaceFoldTheorems.QuantumInfoSandbox',
  'ForkRaceFoldTheorems.ThermoInformationBridge',
] as const;

export type GnosisPhysicsSandboxLeanModule =
  (typeof GNOSIS_PHYSICS_SANDBOX_LEAN_MODULES)[number];

/** Explicit disclaimer string for manifests and UIs. */
export const PHYSICS_SANDBOX_NUMERIC_DISCLAIMER =
  'This TypeScript module is a deterministic numeric toy; it is not checked against the Lean physics sandbox.';

/**
 * IEEE `Math.LN2` — same numeric value as the fair-bit Shannon entropy in nats proved in Lean as
 * `ForkRaceFoldTheorems.ThermoInformationBridge.shannon_fair_binary_fin2` (not a proof obligation here).
 */
export const SHANNON_FAIR_BINARY_NATS_JS = Math.LN2;

/** One explicit Euler step: `x' = f(t, x)` on `ℝ`. */
export function eulerStep(
  t: number,
  x: number,
  dt: number,
  f: (t: number, x: number) => number,
): number {
  return x + dt * f(t, x);
}

/** Midpoint (RK2) step for `x' = f(t, x)` — typically lower drift than Euler for fixed `dt`. */
export function rk2Step(
  t: number,
  x: number,
  dt: number,
  f: (t: number, x: number) => number,
): number {
  const k1 = f(t, x);
  const k2 = f(t + dt / 2, x + (dt / 2) * k1);
  return x + dt * k2;
}

/** Integrate `x' = v` (constant) from `t0` to `t1` with fixed `dt` Euler steps (`x(0)=0`). */
export function integrateConstVelocity(
  v: number,
  t0: number,
  t1: number,
  dt: number,
): { xs: number[]; times: number[] } {
  const xs: number[] = [];
  const times: number[] = [];
  let t = t0;
  let x = 0;
  while (t <= t1 + 1e-12) {
    times.push(t);
    xs.push(x);
    x = eulerStep(t, x, dt, () => v);
    t += dt;
  }
  return { xs, times };
}

/**
 * Same as `integrateLinearAutonomous` but uses `rk2Step` (see `ClassicalSandbox.linearField` in Lean).
 */
export function integrateLinearAutonomousRk2(
  a: number,
  x0: number,
  t0: number,
  t1: number,
  dt: number,
): { xs: number[]; times: number[] } {
  const xs: number[] = [];
  const times: number[] = [];
  let t = t0;
  let x = x0;
  while (t <= t1 + 1e-12) {
    times.push(t);
    xs.push(x);
    x = rk2Step(t, x, dt, (_, xx) => a * xx);
    t += dt;
  }
  return { xs, times };
}

/**
 * Explicit Euler for the autonomous linear ODE `x' = a x` (Lean classical toy: `ClassicalSandbox.linearField`).
 * Closed form: `x(t) = x0 * exp(a * (t - t0))`; this routine is for demos only.
 */
export function integrateLinearAutonomous(
  a: number,
  x0: number,
  t0: number,
  t1: number,
  dt: number,
): { xs: number[]; times: number[] } {
  const xs: number[] = [];
  const times: number[] = [];
  let t = t0;
  let x = x0;
  while (t <= t1 + 1e-12) {
    times.push(t);
    xs.push(x);
    x = eulerStep(t, x, dt, (_, xx) => a * xx);
    t += dt;
  }
  return { xs, times };
}
