// @a0n/gnosis/src/pisot-sieve.ts
/**
 * pisot-sieve.ts -- Pisot-Native Prime Sieve
 *
 * Formalized in Lean at Lean/ForkRaceFoldTheorems/PisotPrimeSieve.lean.
 *
 * A standard sieve crosses out multiples on a linear tape. A Pisot Sieve
 * identifies "Resonance Interference" in the 3D Pisot lattice.
 * Primes are "Bright" vertices that cannot be reached via the lattice product
 * of two smaller, non-unit Pisot vectors.
 */

import { pisotF } from './betti/pisot-primitive';
import { PELL_TOWER, TRIB_TOWER, CLOCK_MAX_INDEX } from './quasicrystalline-clock';

export interface PisotVector {
  phi: number;
  pell: number;
  trib: number;
}

/**
 * Projects a 3D Pisot coordinate down to its scalar Natural Number.
 */
export function projectVector(v: PisotVector): bigint {
  const f = BigInt(pisotF(v.phi));
  const p = PELL_TOWER[v.pell];
  const t = TRIB_TOWER[v.trib];
  return f * p * t;
}

/**
 * A vector is a "unit" if its scalar projection is exactly 1.
 */
export function isUnitVector(v: PisotVector): boolean {
  return projectVector(v) === 1n;
}

/**
 * The Pisot Sieve: Identifies primes up to a maximum scalar value by
 * projecting all valid combinations of the 3D tensor and detecting
 * which values have NO composite "alias" in the lattice.
 */
export function runPisotSieve(maxScalar: number): Set<number> {
  const maxN = BigInt(maxScalar);
  const shadowed = new Set<bigint>();
  
  // To sieve efficiently, we only need to iterate up to the points
  // where the projection doesn't exceed our maximum.
  const vectors: PisotVector[] = [];
  
  for (let i = 0; i <= CLOCK_MAX_INDEX; i++) {
    if (BigInt(pisotF(i)) > maxN) break;
    for (let j = 0; j <= CLOCK_MAX_INDEX; j++) {
      if (BigInt(pisotF(i)) * PELL_TOWER[j] > maxN) break;
      for (let k = 0; k <= CLOCK_MAX_INDEX; k++) {
        const v = { phi: i, pell: j, trib: k };
        const val = projectVector(v);
        if (val > maxN) break;
        
        // We only care about non-unit vectors to form lattice products
        if (!isUnitVector(v)) {
          vectors.push(v);
        }
      }
    }
  }

  // "Phase Clashing" / Sieve mechanism:
  // Any value that can be formed by the product of two non-unit vectors is Composite (Shadowed).
  // In the real system, this is done via FORK (lattice multiplication), but here
  // we do the direct scalar projection check for demonstration.
  for (let a = 0; a < vectors.length; a++) {
    const valA = projectVector(vectors[a]);
    for (let b = a; b < vectors.length; b++) {
      const valB = projectVector(vectors[b]);
      const product = valA * valB;
      if (product <= maxN) {
        shadowed.add(product);
      }
    }
  }

  // Any projection > 1 that hasn't been shadowed is a Pisot-native Prime (Bright).
  const primes = new Set<number>();
  for (const v of vectors) {
    const val = projectVector(v);
    if (val > 1n && !shadowed.has(val)) {
      primes.add(Number(val));
    }
  }

  return primes;
}
