// @a0n/gnosis/src/polyglot-typebridge.ts
/**
 * Cross-language type bridge.
 *
 * Given function signatures extracted by the polyglot scanner, infers
 * type compatibility across language boundaries. When a Python function
 * returns `dict` and the next Go function expects `map[string]interface{}`,
 * the type bridge determines they are compatible (both are JSON objects).
 *
 * When types are incompatible (e.g., Python `list[float]` feeding Go
 * `string`), the bridge flags the mismatch at analysis time, before
 * any code runs.
 */

import type { GraphAST, ASTNode, ASTEdge } from './betty/compiler';

/**
 * A type compatibility check result.
 */
export interface TypeCheckResult {
  /** Whether the types are compatible. */
  compatible: boolean;
  /** Confidence level (0.0-1.0). */
  confidence: number;
  /** Human-readable explanation. */
  explanation: string;
  /** The source type (producer). */
  sourceType: string;
  /** The target type (consumer). */
  targetType: string;
  /** The source language. */
  sourceLanguage: string;
  /** The target language. */
  targetLanguage: string;
}

/**
 * A type mismatch found in a composed topology.
 */
export interface TypeMismatch {
  /** The edge where the mismatch occurs. */
  fromNode: string;
  toNode: string;
  /** The check result with details. */
  check: TypeCheckResult;
}

/**
 * Canonical type categories that map across languages.
 */
type CanonicalType =
  | 'void'
  | 'boolean'
  | 'integer'
  | 'float'
  | 'string'
  | 'bytes'
  | 'array'
  | 'map'
  | 'object'
  | 'error'
  | 'optional'
  | 'future'
  | 'unknown';

/**
 * Map a language-specific type string to a canonical type.
 */
function canonicalize(type: string, language: string): CanonicalType {
  const normalized = type.trim().toLowerCase();

  // Void / None / Unit.
  if (['void', 'none', 'nil', 'null', 'unit', '()'].includes(normalized))
    return 'void';

  // Boolean.
  if (['bool', 'boolean', 'true', 'false'].includes(normalized))
    return 'boolean';

  // Integer.
  if (
    [
      'int',
      'integer',
      'i8',
      'i16',
      'i32',
      'i64',
      'u8',
      'u16',
      'u32',
      'u64',
      'usize',
      'isize',
      'long',
      'short',
      'byte',
      'int8',
      'int16',
      'int32',
      'int64',
      'uint',
      'uint8',
      'uint16',
      'uint32',
      'uint64',
      'number',
    ].some((t) => normalized.startsWith(t) || normalized === t)
  ) {
    if (normalized === 'number')
      return language === 'typescript' ? 'float' : 'integer';
    return 'integer';
  }

  // Float.
  if (
    [
      'float',
      'double',
      'f32',
      'f64',
      'float32',
      'float64',
      'decimal',
      'real',
    ].some((t) => normalized.startsWith(t))
  )
    return 'float';

  // String.
  if (
    ['str', 'string', '&str', 'string!', 'text', 'char'].some((t) =>
      normalized.startsWith(t)
    )
  )
    return 'string';

  // Bytes.
  if (
    [
      'bytes',
      '[]byte',
      'bytearray',
      'uint8array',
      'buffer',
      '[u8]',
      'vec<u8>',
    ].some((t) => normalized.includes(t))
  )
    return 'bytes';

  // Array / List.
  if (
    ['list', 'array', 'vec', '[]', 'slice', 'sequence', 'iterable'].some(
      (t) => normalized.startsWith(t) || normalized.includes(t)
    )
  )
    return 'array';

  // Map / Dict / Object.
  if (
    ['dict', 'map', 'hashmap', 'hash', 'record', 'object'].some(
      (t) => normalized.startsWith(t) || normalized.includes(t)
    )
  )
    return 'map';

  // Error.
  if (
    ['error', 'exception', 'err', 'result'].some((t) =>
      normalized.startsWith(t)
    )
  )
    return 'error';

  // Optional.
  if (
    ['optional', 'option', 'maybe'].some((t) => normalized.startsWith(t)) ||
    normalized.endsWith('?') ||
    normalized.endsWith('| null') ||
    normalized.endsWith('| undefined')
  )
    return 'optional';

  // Future / Promise / Async.
  if (
    ['promise', 'future', 'task', 'async', 'deferred'].some((t) =>
      normalized.includes(t)
    )
  )
    return 'future';

  // Interface / Any / Dynamic.
  if (
    ['interface{}', 'any', 'object', 'dynamic', 'unknown', 'value'].includes(
      normalized
    )
  )
    return 'unknown';

  return 'object';
}

/**
 * Check type compatibility between two language-specific types.
 */
export function checkTypeCompatibility(
  sourceType: string,
  targetType: string,
  sourceLanguage: string,
  targetLanguage: string
): TypeCheckResult {
  const sourceCanonical = canonicalize(sourceType, sourceLanguage);
  const targetCanonical = canonicalize(targetType, targetLanguage);

  // Exact match.
  if (sourceCanonical === targetCanonical) {
    return {
      compatible: true,
      confidence: 1.0,
      explanation: `Both are ${sourceCanonical} types`,
      sourceType,
      targetType,
      sourceLanguage,
      targetLanguage,
    };
  }

  // Unknown/any accepts everything.
  if (targetCanonical === 'unknown' || sourceCanonical === 'unknown') {
    return {
      compatible: true,
      confidence: 0.7,
      explanation: `${
        targetCanonical === 'unknown' ? 'Target' : 'Source'
      } accepts any type`,
      sourceType,
      targetType,
      sourceLanguage,
      targetLanguage,
    };
  }

  // Numeric coercion: int → float is safe.
  if (sourceCanonical === 'integer' && targetCanonical === 'float') {
    return {
      compatible: true,
      confidence: 0.9,
      explanation: 'Integer can widen to float',
      sourceType,
      targetType,
      sourceLanguage,
      targetLanguage,
    };
  }

  // Map and object are interchangeable across languages.
  if (
    (sourceCanonical === 'map' && targetCanonical === 'object') ||
    (sourceCanonical === 'object' && targetCanonical === 'map')
  ) {
    return {
      compatible: true,
      confidence: 0.8,
      explanation: 'Map and object types are structurally compatible via JSON',
      sourceType,
      targetType,
      sourceLanguage,
      targetLanguage,
    };
  }

  // Optional wrapping: T is compatible with Optional<T>.
  if (targetCanonical === 'optional') {
    return {
      compatible: true,
      confidence: 0.8,
      explanation: 'Value can be wrapped in optional',
      sourceType,
      targetType,
      sourceLanguage,
      targetLanguage,
    };
  }

  // Future unwrapping: Promise<T> is compatible with T (await boundary).
  if (sourceCanonical === 'future') {
    return {
      compatible: true,
      confidence: 0.7,
      explanation: 'Async value will be awaited at topology boundary',
      sourceType,
      targetType,
      sourceLanguage,
      targetLanguage,
    };
  }

  // Incompatible.
  return {
    compatible: false,
    confidence: 0.9,
    explanation: `${sourceCanonical} (${sourceLanguage}) is not compatible with ${targetCanonical} (${targetLanguage})`,
    sourceType,
    targetType,
    sourceLanguage,
    targetLanguage,
  };
}

/**
 * Check all edges in a composed topology for type mismatches.
 *
 * Uses function signatures attached to node properties to determine
 * the output type of each producer and input type of each consumer.
 */
export function checkTopologyTypes(
  ast: GraphAST,
  signatures: Map<
    string,
    { returnType?: string; params?: Array<{ type_annotation?: string }> }
  >,
  languages: Map<string, string>
): TypeMismatch[] {
  const mismatches: TypeMismatch[] = [];

  for (const edge of ast.edges) {
    if (edge.type !== 'PROCESS') continue; // Only check sequential data flow edges.

    for (const sourceId of edge.sourceIds) {
      for (const targetId of edge.targetIds) {
        const sourceSig = signatures.get(sourceId.trim());
        const targetSig = signatures.get(targetId.trim());
        const sourceLang = languages.get(sourceId.trim()) ?? 'unknown';
        const targetLang = languages.get(targetId.trim()) ?? 'unknown';

        if (
          !sourceSig?.returnType ||
          !targetSig?.params?.[0]?.type_annotation
        ) {
          continue; // Can't check without type annotations.
        }

        const check = checkTypeCompatibility(
          sourceSig.returnType,
          targetSig.params[0].type_annotation,
          sourceLang,
          targetLang
        );

        if (!check.compatible) {
          mismatches.push({
            fromNode: sourceId.trim(),
            toNode: targetId.trim(),
            check,
          });
        }
      }
    }
  }

  return mismatches;
}
