// @a0n/gnosis/src/polyglot-scan-cache.ts
/**
 * Polyglot Scan Cache
 *
 * Caches polyglot scanner results per project to avoid re-scanning
 * unchanged code. Cache key is a content hash of all source files
 * in the project.
 *
 * Cache location: .gnosis/cache/polyglot-scan/{project-hash}.json
 *
 * @module @a0n/gnosis/polyglot-scan-cache
 */

import { createHash } from 'crypto';
import {
  existsSync,
  mkdirSync,
  readFileSync,
  readdirSync,
  statSync,
  unlinkSync,
  writeFileSync,
} from 'fs';
import { extname, join, relative } from 'path';
import type { ProjectScanResult } from './polyglot-checkpoint-bridge';

// ═══════════════════════════════════════════════════════════════════════
// Types
// ═══════════════════════════════════════════════════════════════════════

interface CacheEntry {
  /** Content hash of all source files at scan time. */
  contentHash: string;
  /** Timestamp of scan. */
  scannedAt: string;
  /** The cached scan result. */
  scanResult: ProjectScanResult;
}

// ═══════════════════════════════════════════════════════════════════════
// Configuration
// ═══════════════════════════════════════════════════════════════════════

const CACHE_DIR_NAME = '.gnosis';
const CACHE_SUBDIR = 'cache/polyglot-scan';

const IGNORED_DIRS = new Set([
  'node_modules',
  '.next',
  'dist',
  'build',
  '.git',
  'coverage',
  '.wrangler',
  '.a0',
  '.gnosis',
  '.lake',
  '.tmp',
  'target',
  '.turbo',
]);

const SOURCE_EXTENSIONS = new Set([
  '.ts',
  '.tsx',
  '.js',
  '.jsx',
  '.py',
  '.rs',
  '.go',
  '.rb',
  '.java',
  '.kt',
  '.swift',
  '.c',
  '.cpp',
  '.h',
  '.hpp',
]);

// ═══════════════════════════════════════════════════════════════════════
// Cache Directory
// ═══════════════════════════════════════════════════════════════════════

function ensureCacheDir(workspaceRoot: string): string {
  const cacheDir = join(workspaceRoot, CACHE_DIR_NAME, CACHE_SUBDIR);
  if (!existsSync(cacheDir)) {
    mkdirSync(cacheDir, { recursive: true });
  }
  return cacheDir;
}

// ═══════════════════════════════════════════════════════════════════════
// Content Hashing
// ═══════════════════════════════════════════════════════════════════════

/**
 * Compute a content hash for all source files in a project directory.
 * The hash changes when any source file is added, removed, or modified.
 */
export function computeProjectHash(projectRoot: string): string {
  const hash = createHash('sha256');
  const files = collectSourceFiles(projectRoot);

  // Sort for deterministic ordering
  files.sort();

  for (const file of files) {
    const relPath = relative(projectRoot, file);
    hash.update(relPath);
    try {
      const content = readFileSync(file);
      hash.update(content);
    } catch {
      // File disappeared between listing and reading — hash the path only
      hash.update('__missing__');
    }
  }

  return hash.digest('hex').slice(0, 16);
}

function collectSourceFiles(root: string): string[] {
  const files: string[] = [];
  const queue = [root];

  while (queue.length > 0) {
    const dir = queue.pop()!;
    let entries;
    try {
      entries = readdirSync(dir, { withFileTypes: true });
    } catch {
      continue;
    }

    for (const entry of entries) {
      if (entry.isDirectory()) {
        if (!IGNORED_DIRS.has(entry.name)) queue.push(join(dir, entry.name));
      } else if (SOURCE_EXTENSIONS.has(extname(entry.name))) {
        files.push(join(dir, entry.name));
      }
    }
  }

  return files;
}

// ═══════════════════════════════════════════════════════════════════════
// Cache Key
// ═══════════════════════════════════════════════════════════════════════

function cacheKeyForProject(projectName: string): string {
  // Sanitize project name for filesystem
  return projectName.replace(/[^a-zA-Z0-9._-]/g, '_');
}

// ═══════════════════════════════════════════════════════════════════════
// Public API
// ═══════════════════════════════════════════════════════════════════════

/**
 * Try to load a cached scan result for a project.
 * Returns null if cache miss or stale.
 */
export function loadCachedScan(
  workspaceRoot: string,
  projectName: string,
  projectRoot: string,
): ProjectScanResult | null {
  const cacheDir = ensureCacheDir(workspaceRoot);
  const cacheFile = join(cacheDir, `${cacheKeyForProject(projectName)}.json`);

  if (!existsSync(cacheFile)) return null;

  try {
    const raw = readFileSync(cacheFile, 'utf8');
    const entry: CacheEntry = JSON.parse(raw);

    // Check if content hash still matches
    const currentHash = computeProjectHash(projectRoot);
    if (entry.contentHash !== currentHash) return null;

    return entry.scanResult;
  } catch {
    return null;
  }
}

/**
 * Save a scan result to cache.
 */
export function saveScanToCache(
  workspaceRoot: string,
  projectName: string,
  projectRoot: string,
  scanResult: ProjectScanResult,
): void {
  const cacheDir = ensureCacheDir(workspaceRoot);
  const cacheFile = join(cacheDir, `${cacheKeyForProject(projectName)}.json`);

  const entry: CacheEntry = {
    contentHash: computeProjectHash(projectRoot),
    scannedAt: new Date().toISOString(),
    scanResult,
  };

  try {
    writeFileSync(cacheFile, JSON.stringify(entry, null, 2), 'utf8');
  } catch {
    // Cache write failure is non-fatal
  }
}

/**
 * Invalidate the cache for a specific project.
 */
export function invalidateCache(
  workspaceRoot: string,
  projectName: string,
): void {
  const cacheDir = join(workspaceRoot, CACHE_DIR_NAME, CACHE_SUBDIR);
  const cacheFile = join(cacheDir, `${cacheKeyForProject(projectName)}.json`);

  try {
    unlinkSync(cacheFile);
  } catch {
    // Already missing or not writable — fine
  }
}

/**
 * Invalidate all cached scan results.
 */
export function invalidateAllCaches(workspaceRoot: string): void {
  const cacheDir = join(workspaceRoot, CACHE_DIR_NAME, CACHE_SUBDIR);
  if (!existsSync(cacheDir)) return;

  try {
    const entries = readdirSync(cacheDir);
    for (const entry of entries) {
      if (entry.endsWith('.json')) {
        try {
          unlinkSync(join(cacheDir, entry));
        } catch {
          // Best effort
        }
      }
    }
  } catch {
    // Non-fatal
  }
}
