// @a0n/gnosis/src/polyglot-trace.ts
/**
 * Cross-language execution trace for polyglot topologies.
 *
 * When a polyglot function fails, this module reconstructs the full
 * execution path through the GG topology, mapping errors back to their
 * source locations across language boundaries.
 *
 * Example output:
 *
 *   Error in node 'ml_predict' (python:ml.py:42)
 *     ← fed by node 'fetch_data' (go:api.go:108)
 *     ← fed by node 'parse_input' (rust:parser.rs:23)
 *
 *   Topology path: entry →[PROCESS]→ parse_input →[PROCESS]→ fetch_data →[PROCESS]→ ml_predict
 *   Betty analysis: no deficit on this path, error is semantic (not topological)
 */

import type { GraphAST } from './betty/compiler';
import type { MappedError } from './runtime/subprocess-invoker';
import type { PolyglotExecutionManifest } from './polyglot-execution-bridge';

/**
 * A single frame in a cross-language stack trace.
 */
export interface PolyglotTraceFrame {
  /** GG topology node ID. */
  nodeId: string;
  /** Node labels (e.g., ["Statement", "PolyglotBridgeCall"]). */
  labels: string[];
  /** Language of this node's execution. */
  language: string;
  /** Source file path. */
  filePath: string;
  /** Function/callee name. */
  functionName: string;
  /** Source line (if available from error mapping). */
  line?: number;
  /** Source column (if available from error mapping). */
  column?: number;
  /** Edge type that led to this node (PROCESS, FORK, FOLD, etc.). */
  incomingEdgeType?: string;
  /** Execution status. */
  status: 'completed' | 'error' | 'skipped' | 'pending';
  /** Duration in ms (if available). */
  durationMs?: number;
}

/**
 * Complete cross-language execution trace.
 */
export interface PolyglotExecutionTrace {
  /** Ordered frames from entry to error (or completion). */
  frames: PolyglotTraceFrame[];
  /** The node that failed (if any). */
  failedNode?: string;
  /** The mapped error with language-specific context. */
  mappedError?: MappedError;
  /** Topology-level analysis. */
  analysis: {
    /** Total nodes in the path. */
    pathLength: number;
    /** Languages involved in the path. */
    languages: string[];
    /** Whether the error is topological (deficit/deadlock) or semantic. */
    errorKind: 'topological' | 'semantic' | 'none';
    /** Edge types traversed. */
    edgeTypes: string[];
  };
}

/**
 * Build a reverse adjacency map for the topology.
 * Maps each node ID to the set of (sourceNodeId, edgeType) that feed into it.
 */
function buildReverseAdjacency(
  ast: GraphAST
): Map<string, Array<{ sourceId: string; edgeType: string }>> {
  const reverse = new Map<
    string,
    Array<{ sourceId: string; edgeType: string }>
  >();

  for (const edge of ast.edges) {
    for (const targetId of edge.targetIds) {
      const trimmed = targetId.trim();
      if (!reverse.has(trimmed)) {
        reverse.set(trimmed, []);
      }
      for (const sourceId of edge.sourceIds) {
        reverse.get(trimmed)!.push({
          sourceId: sourceId.trim(),
          edgeType: edge.type,
        });
      }
    }
  }

  return reverse;
}

/**
 * Trace backward from a failed node to reconstruct the execution path.
 */
function traceBackward(
  ast: GraphAST,
  failedNodeId: string,
  reverseAdj: Map<string, Array<{ sourceId: string; edgeType: string }>>,
  manifest: PolyglotExecutionManifest
): PolyglotTraceFrame[] {
  const plansByNodeId = new Map(
    manifest.node_execution_plans.map((p) => [p.node_id, p] as const)
  );

  const frames: PolyglotTraceFrame[] = [];
  const visited = new Set<string>();
  let currentId: string | undefined = failedNodeId;

  while (currentId && !visited.has(currentId)) {
    visited.add(currentId);
    const node = ast.nodes.get(currentId);
    if (!node) break;

    const plan = plansByNodeId.get(currentId);
    const predecessors = reverseAdj.get(currentId);
    const incomingEdge = predecessors?.[0];

    const frame: PolyglotTraceFrame = {
      nodeId: currentId,
      labels: node.labels,
      language: node.properties.language ?? manifest.language,
      filePath: node.properties.file ?? manifest.file_path,
      functionName:
        plan?.callee ??
        node.properties.callee ??
        node.properties.name ??
        currentId,
      status: currentId === failedNodeId ? 'error' : 'completed',
    };
    if (incomingEdge?.edgeType !== undefined) {
      frame.incomingEdgeType = incomingEdge.edgeType;
    }
    frames.unshift(frame);

    // Walk backward to the first predecessor.
    currentId = incomingEdge?.sourceId;
  }

  return frames;
}

/**
 * Build a cross-language execution trace from a failed polyglot execution.
 *
 * @param ast The GraphAST topology that was being executed.
 * @param failedNodeId The node ID where the error occurred.
 * @param mappedError The language-aware error mapping from the subprocess invoker.
 * @param manifest The polyglot execution manifest.
 * @param executionLog The engine execution logs (optional, for enrichment).
 */
export function buildExecutionTrace(
  ast: GraphAST,
  failedNodeId: string,
  mappedError: MappedError | undefined,
  manifest: PolyglotExecutionManifest,
  _executionLog?: string
): PolyglotExecutionTrace {
  const reverseAdj = buildReverseAdjacency(ast);
  const frames = traceBackward(ast, failedNodeId, reverseAdj, manifest);

  // Enrich the failed frame with mapped error details.
  if (mappedError && frames.length > 0) {
    const failedFrame = frames[frames.length - 1]!;
    if (mappedError.line !== undefined) {
      failedFrame.line = mappedError.line;
    }
    if (mappedError.column !== undefined) {
      failedFrame.column = mappedError.column;
    }
    failedFrame.language = mappedError.language;
    failedFrame.filePath = mappedError.filePath;
  }

  // Analyze the path.
  const languages = [...new Set(frames.map((f) => f.language))];
  const edgeTypes = frames
    .map((f) => f.incomingEdgeType)
    .filter((t): t is string => t !== undefined);

  // Determine if the error is topological or semantic.
  const hasVent = edgeTypes.includes('VENT');
  const hasInterfere = edgeTypes.includes('SLIVER');
  const errorKind =
    hasVent || hasInterfere
      ? 'topological'
      : failedNodeId
      ? 'semantic'
      : 'none';

  const trace: PolyglotExecutionTrace = {
    frames,
    failedNode: failedNodeId,
    analysis: {
      pathLength: frames.length,
      languages,
      errorKind,
      edgeTypes,
    },
  };
  if (mappedError !== undefined) {
    trace.mappedError = mappedError;
  }
  return trace;
}

/**
 * Format a cross-language execution trace as a human-readable string.
 */
export function formatExecutionTrace(trace: PolyglotExecutionTrace): string {
  const lines: string[] = [];

  if (trace.mappedError) {
    const loc = trace.mappedError.line
      ? `:${trace.mappedError.line}${
          trace.mappedError.column ? `:${trace.mappedError.column}` : ''
        }`
      : '';
    lines.push(
      `Error in node '${trace.failedNode}' (${trace.mappedError.language}:${trace.mappedError.filePath}${loc})`
    );
    lines.push(`  ${trace.mappedError.message}`);
    lines.push('');
  }

  // Reverse trace: show the chain from error back to entry.
  if (trace.frames.length > 1) {
    lines.push('Execution path:');
    for (let i = trace.frames.length - 1; i >= 0; i--) {
      const frame = trace.frames[i]!;
      const loc = frame.line ? `:${frame.line}` : '';
      const prefix = i === trace.frames.length - 1 ? '  ERROR ' : '      ← ';
      const edge = frame.incomingEdgeType
        ? ` ←[${frame.incomingEdgeType}]`
        : '';
      lines.push(
        `${prefix}${frame.functionName} (${frame.language}:${frame.filePath}${loc})${edge}`
      );
    }
    lines.push('');
  }

  // Topology path summary.
  const pathStr = trace.frames
    .map((f, i) => {
      const edge = f.incomingEdgeType ? ` →[${f.incomingEdgeType}]→ ` : '';
      return i === 0 ? f.nodeId : `${edge}${f.nodeId}`;
    })
    .join('');
  lines.push(`Topology path: ${pathStr}`);

  // Analysis.
  lines.push(
    `Languages: ${trace.analysis.languages.join(', ')} | ` +
      `Path length: ${trace.analysis.pathLength} | ` +
      `Error kind: ${trace.analysis.errorKind}`
  );

  return lines.join('\n');
}
