// @a0n/gnosis/src/polyglot-compose.ts
/**
 * gnode compose -- automatic cross-language program composition.
 *
 * Takes source files in different languages, parses each through the polyglot
 * scanner, analyzes function signatures and data flow, and generates a .gg
 * topology that wires them together. Betty verifies the composed program.
 *
 * Also supports:
 * - Translation: source → GG → target language
 * - Language fitness: find the optimal language for each function
 * - Topo-race: race multiple language implementations, fastest correct wins
 * - Least bule: minimize beta cost across the topology
 */

import type { GraphAST, ASTNode, ASTEdge } from './betty/compiler';
import type { PolyglotExecutionManifest } from './polyglot-execution-bridge';
import {
  scaffoldTopology,
  type ScaffoldAssignment,
} from './polyglot-scaffold';
import {
  denoteType,
  checkTypeCompatibility,
  type TopologyType,
} from './betty/semantic-compatibility';

// ─── Types ───────────────────────────────────────────────────────────────────

export { scaffoldTopology };
export type { ScaffoldAssignment };

/**
 * Extracted function signature from polyglot analysis.
 */
export interface ExtractedFunction {
  /** Source file path. */
  filePath: string;
  /** Detected language. */
  language: string;
  /** Function name. */
  name: string;
  /** Parameter names (from source analysis). */
  params: string[];
  /** Functions called by this function. */
  callees: string[];
  /** The GG topology for this function. */
  ast: GraphAST;
  /** The .gg source text. */
  ggSource: string;
  /** Number of nodes in the topology (complexity proxy). */
  nodeCount: number;
  /** Edge types present (FORK, RACE, FOLD, etc.). */
  edgeTypes: string[];
  /** Topology-level semantic return type (from language-specific annotation). */
  semanticReturnType?: TopologyType;
  /** Topology-level semantic param types. */
  semanticParamTypes?: TopologyType[];
}

/**
 * A composed topology connecting functions across files/languages.
 */
export interface ComposedTopology {
  /** The generated .gg source. */
  ggSource: string;
  /** All functions included in the composition. */
  functions: ExtractedFunction[];
  /** Inferred connections between functions. */
  connections: FunctionConnection[];
  /** The entry point function. */
  entryFunction: string;
  /** Languages involved. */
  languages: string[];
  /** Execution manifest for the composed topology. */
  manifest: PolyglotExecutionManifest;
}

/**
 * An inferred connection between two functions.
 */
export interface FunctionConnection {
  /** The calling function. */
  from: { file: string; function: string; language: string };
  /** The called function. */
  to: { file: string; function: string; language: string };
  /** How the connection was inferred. */
  inference: 'direct-call' | 'name-match' | 'type-compatible' | 'sequential';
  /** Edge type in the GG topology. */
  edgeType: 'PROCESS' | 'FORK' | 'RACE';
}

/**
 * Language fitness score for a function.
 */
export interface LanguageFitness {
  language: string;
  /** Overall fitness score (0.0 - 1.0, higher = better fit). */
  score: number;
  /** Breakdown by category. */
  breakdown: {
    /** How well the language handles this function's compute pattern. */
    computeFit: number;
    /** How well the language handles this function's IO pattern. */
    ioFit: number;
    /** How well the language handles this function's concurrency pattern. */
    concurrencyFit: number;
    /** Raw performance tier (lower = faster). */
    performanceTier: number;
    /** Ecosystem maturity for this problem domain. */
    ecosystemFit: number;
  };
  /** Human-readable rationale. */
  rationale: string;
}

/**
 * Topo-race result: multiple language implementations raced.
 */
export interface TopoRaceResult {
  /** The winning language. */
  winner: string;
  /** All race entries with their results. */
  entries: Array<{
    language: string;
    durationMs: number;
    status: 'success' | 'error' | 'timeout';
    value?: unknown;
  }>;
  /** The optimal .gg topology using the winner. */
  optimalTopology: string;
}

// ─── Function Extraction ─────────────────────────────────────────────────────

/**
 * Extract function information from a polyglot analysis result.
 */
export function extractFunctions(
  analysisResult: {
    functions: Array<{
      functionName: string;
      ast: GraphAST;
      ggSource: string;
      /** Signature from the polyglot scanner (includes semantic contract). */
      signature?: {
        params: Array<{
          name: string;
          type_annotation: string | null;
          semantic_type?: unknown;
        }>;
        return_type: string | null;
        callees: string[];
        semantic_contract?: {
          param_types: unknown[];
          return_type: unknown;
          predicates: unknown[];
        };
      };
    }>;
    language: string;
    errors: string[];
  },
  filePath: string
): ExtractedFunction[] {
  return analysisResult.functions.map((func) => {
    // Extract callees from PolyglotBridgeCall nodes.
    const callees: string[] = [];
    const params: string[] = [];

    for (const [, node] of func.ast.nodes) {
      if (node.properties.callee) {
        callees.push(node.properties.callee);
      }
    }

    // Use signature callees if available (more complete than AST extraction).
    if (func.signature?.callees) {
      for (const c of func.signature.callees) {
        if (!callees.includes(c)) callees.push(c);
      }
    }

    // Extract param names from signature.
    if (func.signature?.params) {
      for (const p of func.signature.params) {
        params.push(p.name);
      }
    }

    const edgeTypes = [...new Set(func.ast.edges.map((e) => e.type))];

    // Extract semantic types from the scanner's semantic contract.
    let semanticReturnType: TopologyType | undefined;
    let semanticParamTypes: TopologyType[] | undefined;

    if (func.signature?.semantic_contract) {
      const contract = func.signature.semantic_contract;
      semanticReturnType = contract.return_type as TopologyType | undefined;
      if (contract.param_types && contract.param_types.length > 0) {
        semanticParamTypes = contract.param_types as TopologyType[];
      }
    } else if (func.signature?.return_type) {
      // Fallback: compute denotation from type string on the TS side.
      semanticReturnType = denoteType(
        analysisResult.language,
        func.signature.return_type
      );
      if (func.signature.params) {
        semanticParamTypes = func.signature.params
          .filter((p) => p.type_annotation)
          .map((p) => denoteType(analysisResult.language, p.type_annotation!));
      }
    }

    return {
      filePath,
      language: analysisResult.language,
      name: func.functionName,
      params,
      callees,
      ast: func.ast,
      ggSource: func.ggSource,
      nodeCount: func.ast.nodes.size,
      edgeTypes,
      ...(semanticReturnType !== undefined ? { semanticReturnType } : {}),
      ...(semanticParamTypes !== undefined ? { semanticParamTypes } : {}),
    };
  });
}

// ─── Composition Engine ──────────────────────────────────────────────────────

/**
 * Infer connections between functions across files.
 *
 * Strategies:
 * 1. Direct call: function A calls function B by name.
 * 2. Name match: function names follow a pipeline pattern (parse → transform → output).
 * 3. Sequential: if no other signal, chain functions in file order.
 */
function inferConnections(
  functions: ExtractedFunction[]
): FunctionConnection[] {
  const connections: FunctionConnection[] = [];
  const functionsByName = new Map(functions.map((f) => [f.name, f]));

  // Strategy 1: Direct call matching.
  for (const func of functions) {
    for (const callee of func.callees) {
      const target = functionsByName.get(callee);
      if (target && target.filePath !== func.filePath) {
        connections.push({
          from: {
            file: func.filePath,
            function: func.name,
            language: func.language,
          },
          to: {
            file: target.filePath,
            function: target.name,
            language: target.language,
          },
          inference: 'direct-call',
          edgeType: 'PROCESS',
        });
      }
    }
  }

  // Strategy 1.5: Type-compatible matching.
  // If a function's return type is compatible with another function's param type,
  // and they are in different files, create a PROCESS edge.
  if (connections.length === 0) {
    for (const source of functions) {
      if (
        !source.semanticReturnType ||
        source.semanticReturnType.kind === 'unknown'
      )
        continue;
      for (const target of functions) {
        if (source === target || source.filePath === target.filePath) continue;
        if (
          !target.semanticParamTypes ||
          target.semanticParamTypes.length === 0
        )
          continue;
        const compat = checkTypeCompatibility(
          source.semanticReturnType,
          target.semanticParamTypes[0]!
        );
        if (
          compat.status === 'compatible' ||
          compat.status === 'proof-obligation'
        ) {
          connections.push({
            from: {
              file: source.filePath,
              function: source.name,
              language: source.language,
            },
            to: {
              file: target.filePath,
              function: target.name,
              language: target.language,
            },
            inference: 'type-compatible',
            edgeType: 'PROCESS',
          });
        }
      }
    }
  }

  // If no direct calls found, use name-based pipeline detection.
  if (connections.length === 0) {
    // Look for pipeline patterns: input/parse/ingest → process/transform → output/serve/render
    const stages = [
      { pattern: /^(input|parse|ingest|read|fetch|load|extract)/i, order: 0 },
      {
        pattern: /^(process|transform|compute|analyze|normalize|convert)/i,
        order: 1,
      },
      {
        pattern: /^(output|serve|render|write|save|store|publish|send)/i,
        order: 2,
      },
    ];

    const staged = functions
      .map((f) => {
        const match = stages.find((s) => s.pattern.test(f.name));
        return { func: f, order: match?.order ?? 1 };
      })
      .sort((a, b) => a.order - b.order);

    for (let i = 0; i < staged.length - 1; i++) {
      const from = staged[i]!.func;
      const to = staged[i + 1]!.func;
      if (from.filePath !== to.filePath || from.language !== to.language) {
        connections.push({
          from: {
            file: from.filePath,
            function: from.name,
            language: from.language,
          },
          to: { file: to.filePath, function: to.name, language: to.language },
          inference: 'name-match',
          edgeType: 'PROCESS',
        });
      }
    }
  }

  // Fallback: sequential chaining by file order.
  if (connections.length === 0 && functions.length >= 2) {
    // Group by file, take the "main" or first function from each.
    const byFile = new Map<string, ExtractedFunction>();
    for (const func of functions) {
      if (!byFile.has(func.filePath)) {
        // Prefer "main" function, otherwise first.
        const main = functions.find(
          (f) => f.filePath === func.filePath && f.name === 'main'
        );
        byFile.set(func.filePath, main ?? func);
      }
    }

    const ordered = [...byFile.values()];
    for (let i = 0; i < ordered.length - 1; i++) {
      connections.push({
        from: {
          file: ordered[i]!.filePath,
          function: ordered[i]!.name,
          language: ordered[i]!.language,
        },
        to: {
          file: ordered[i + 1]!.filePath,
          function: ordered[i + 1]!.name,
          language: ordered[i + 1]!.language,
        },
        inference: 'sequential',
        edgeType: 'PROCESS',
      });
    }
  }

  return connections;
}

/**
 * Generate a composed .gg topology from functions and their connections.
 */
function generateComposedTopology(
  functions: ExtractedFunction[],
  connections: FunctionConnection[]
): string {
  const lines: string[] = [];
  const languages = [...new Set(functions.map((f) => f.language))];

  lines.push(`// Composed topology: ${languages.join(' + ')}`);
  lines.push(`// Generated by gnode compose`);
  lines.push(
    `// Functions: ${functions
      .map((f) => `${f.name} (${f.language})`)
      .join(', ')}`
  );
  lines.push('');

  // Emit entry node.
  lines.push(`(compose_entry: Entry { name='compose_entry' })`);
  lines.push('');

  // Emit a node per function.
  for (const func of functions) {
    const ext = languageExtension(func.language);
    const fileName = `${toSnakeCase(func.name)}.${ext}`;
    const semanticProps = func.semanticReturnType
      ? `, semantic_return_type='${JSON.stringify(func.semanticReturnType)}'`
      : '';
    lines.push(
      `(${sanitizeId(func.name)}: Statement:PolyglotBridgeCall { ` +
        `language='${func.language}', ` +
        `file='${fileName}', ` +
        `callee='${func.name}', ` +
        `source_file='${func.filePath}'${semanticProps} })`
    );
  }
  lines.push('');

  // Emit return node.
  lines.push(`(compose_return: Return:PolyglotBridgeReturn)`);
  lines.push('');

  // Emit edges.
  // Entry → first function.
  if (functions.length > 0) {
    const first =
      connections.length > 0
        ? connections[0]!.from.function
        : functions[0]!.name;
    lines.push(`(compose_entry)-[:PROCESS]->(${sanitizeId(first)})`);
  }

  // Connection edges.
  for (const conn of connections) {
    lines.push(
      `(${sanitizeId(conn.from.function)})-[:${conn.edgeType}]->(${sanitizeId(
        conn.to.function
      )})`
    );
  }

  // Last function → return.
  if (functions.length > 0) {
    const last =
      connections.length > 0
        ? connections[connections.length - 1]!.to.function
        : functions[functions.length - 1]!.name;
    lines.push(`(${sanitizeId(last)})-[:PROCESS]->(compose_return)`);
  }

  lines.push('');
  return lines.join('\n');
}

/**
 * Build an execution manifest for a composed topology.
 */
function buildComposedManifest(
  functions: ExtractedFunction[],
  _connections: FunctionConnection[]
): PolyglotExecutionManifest {
  const plans: Array<{
    node_id: string;
    source_range: { start_byte: number; end_byte: number };
    kind: 'entry' | 'call' | 'statement' | 'return';
    callee?: string;
  }> = functions.map((func) => ({
    node_id: sanitizeId(func.name),
    source_range: { start_byte: 0, end_byte: 0 },
    kind: 'call',
    callee: func.name,
  }));

  // Add entry and return plans.
  plans.unshift({
    node_id: 'compose_entry',
    source_range: { start_byte: 0, end_byte: 0 },
    kind: 'entry',
  });
  plans.push({
    node_id: 'compose_return',
    source_range: { start_byte: 0, end_byte: 0 },
    kind: 'return',
  });

  return {
    language: functions[0]?.language ?? 'unknown',
    file_path: functions[0]?.filePath ?? '',
    entry_function: 'compose_entry',
    node_execution_plans: plans,
  };
}

/**
 * Compose multiple source files into a single polyglot topology.
 *
 * @param analysisResults Array of polyglot analysis results (one per file).
 * @param filePaths The original file paths (parallel to analysisResults).
 */
export function compose(
  analysisResults: Array<{
    functions: Array<{
      functionName: string;
      ast: GraphAST;
      ggSource: string;
    }>;
    language: string;
    errors: string[];
  }>,
  filePaths: string[]
): ComposedTopology {
  // Extract all functions.
  const allFunctions: ExtractedFunction[] = [];
  for (let i = 0; i < analysisResults.length; i++) {
    const funcs = extractFunctions(analysisResults[i]!, filePaths[i]!);
    allFunctions.push(...funcs);
  }

  // Infer connections.
  const connections = inferConnections(allFunctions);

  // Generate topology.
  const ggSource = generateComposedTopology(allFunctions, connections);
  const manifest = buildComposedManifest(allFunctions, connections);
  const languages = [...new Set(allFunctions.map((f) => f.language))];

  return {
    ggSource,
    functions: allFunctions,
    connections,
    entryFunction: allFunctions[0]?.name ?? 'main',
    languages,
    manifest,
  };
}

// ─── Translation ─────────────────────────────────────────────────────────────

/**
 * Translate a source file to a target language.
 *
 * Pipeline: source → polyglot scanner → GG topology → scaffold in target language.
 * The topology acts as the universal intermediate representation.
 *
 * @param analysisResult The polyglot analysis of the source file.
 * @param sourceFilePath Original file path.
 * @param targetLanguage The language to translate to.
 */
export function translate(
  analysisResult: {
    functions: Array<{
      functionName: string;
      ast: GraphAST;
      ggSource: string;
    }>;
    language: string;
    errors: string[];
  },
  sourceFilePath: string,
  targetLanguage: string
): { files: Array<{ fileName: string; source: string }>; ggSource: string } {
  const functions = extractFunctions(analysisResult, sourceFilePath);

  if (functions.length === 0) {
    return { files: [], ggSource: '' };
  }

  // Use the first function's AST for scaffolding.
  const assignments: ScaffoldAssignment[] = functions.map((func) => ({
    nodeId: sanitizeId(func.name),
    language: targetLanguage,
    functionName: func.name,
  }));

  // Build a combined AST with all function nodes.
  const combinedNodes = new Map<string, ASTNode>();
  const combinedEdges: ASTEdge[] = [];

  for (const func of functions) {
    // Create a node for this function.
    combinedNodes.set(sanitizeId(func.name), {
      id: sanitizeId(func.name),
      labels: ['Statement', 'PolyglotBridgeCall'],
      properties: {
        callee: func.name,
        language: analysisResult.language,
        source_file: sourceFilePath,
        node_count: String(func.nodeCount),
        edge_types: func.edgeTypes.join(','),
      },
    });
  }

  const combinedAst: GraphAST = { nodes: combinedNodes, edges: combinedEdges };

  // Generate the .gg source for the translation.
  const ggLines = [
    `// Translated from ${analysisResult.language} to ${targetLanguage}`,
    `// Source: ${sourceFilePath}`,
    `// Functions: ${functions.map((f) => f.name).join(', ')}`,
    '',
  ];

  for (const func of functions) {
    ggLines.push(`// --- ${func.name} ---`);
    ggLines.push(func.ggSource);
  }

  const ggSource = ggLines.join('\n');

  // Scaffold in target language.
  const result = scaffoldTopology(combinedAst, assignments, ggSource);

  return {
    files: result.files.map((f) => ({
      fileName: f.fileName,
      source: f.source,
    })),
    ggSource,
  };
}

// ─── Language Fitness Analysis ───────────────────────────────────────────────

/**
 * Language characteristics for fitness scoring.
 */
const LANGUAGE_PROFILES: Record<
  string,
  {
    performanceTier: number; // 1=fastest, 5=slowest
    concurrencyModel: string; // goroutines, threads, async, GIL, etc.
    computeStrength: number; // 0-1: numeric/compute-heavy workloads
    ioStrength: number; // 0-1: IO-heavy workloads
    stringStrength: number; // 0-1: string manipulation
    mlEcosystem: number; // 0-1: ML/data science ecosystem
    webEcosystem: number; // 0-1: web/API ecosystem
    systemsEcosystem: number; // 0-1: systems programming ecosystem
  }
> = {
  python: {
    performanceTier: 5,
    concurrencyModel: 'GIL-async',
    computeStrength: 0.3,
    ioStrength: 0.7,
    stringStrength: 0.9,
    mlEcosystem: 1.0,
    webEcosystem: 0.7,
    systemsEcosystem: 0.1,
  },
  rust: {
    performanceTier: 1,
    concurrencyModel: 'ownership-threads',
    computeStrength: 1.0,
    ioStrength: 0.8,
    stringStrength: 0.5,
    mlEcosystem: 0.3,
    webEcosystem: 0.5,
    systemsEcosystem: 1.0,
  },
  go: {
    performanceTier: 2,
    concurrencyModel: 'goroutines',
    computeStrength: 0.7,
    ioStrength: 0.9,
    stringStrength: 0.6,
    mlEcosystem: 0.2,
    webEcosystem: 0.9,
    systemsEcosystem: 0.8,
  },
  javascript: {
    performanceTier: 4,
    concurrencyModel: 'event-loop',
    computeStrength: 0.3,
    ioStrength: 0.9,
    stringStrength: 0.8,
    mlEcosystem: 0.4,
    webEcosystem: 1.0,
    systemsEcosystem: 0.1,
  },
  typescript: {
    performanceTier: 4,
    concurrencyModel: 'event-loop',
    computeStrength: 0.3,
    ioStrength: 0.9,
    stringStrength: 0.8,
    mlEcosystem: 0.4,
    webEcosystem: 1.0,
    systemsEcosystem: 0.1,
  },
  java: {
    performanceTier: 3,
    concurrencyModel: 'threads',
    computeStrength: 0.7,
    ioStrength: 0.7,
    stringStrength: 0.7,
    mlEcosystem: 0.5,
    webEcosystem: 0.8,
    systemsEcosystem: 0.5,
  },
  c: {
    performanceTier: 1,
    concurrencyModel: 'manual-threads',
    computeStrength: 1.0,
    ioStrength: 0.5,
    stringStrength: 0.2,
    mlEcosystem: 0.1,
    webEcosystem: 0.1,
    systemsEcosystem: 1.0,
  },
  cpp: {
    performanceTier: 1,
    concurrencyModel: 'threads',
    computeStrength: 1.0,
    ioStrength: 0.6,
    stringStrength: 0.4,
    mlEcosystem: 0.4,
    webEcosystem: 0.3,
    systemsEcosystem: 1.0,
  },
  ruby: {
    performanceTier: 5,
    concurrencyModel: 'GIL-threads',
    computeStrength: 0.2,
    ioStrength: 0.6,
    stringStrength: 0.9,
    mlEcosystem: 0.2,
    webEcosystem: 0.8,
    systemsEcosystem: 0.1,
  },
  lua: {
    performanceTier: 4,
    concurrencyModel: 'coroutines',
    computeStrength: 0.4,
    ioStrength: 0.5,
    stringStrength: 0.7,
    mlEcosystem: 0.1,
    webEcosystem: 0.2,
    systemsEcosystem: 0.3,
  },
  swift: {
    performanceTier: 2,
    concurrencyModel: 'structured-concurrency',
    computeStrength: 0.8,
    ioStrength: 0.7,
    stringStrength: 0.7,
    mlEcosystem: 0.5,
    webEcosystem: 0.4,
    systemsEcosystem: 0.7,
  },
  kotlin: {
    performanceTier: 3,
    concurrencyModel: 'coroutines',
    computeStrength: 0.6,
    ioStrength: 0.8,
    stringStrength: 0.7,
    mlEcosystem: 0.3,
    webEcosystem: 0.7,
    systemsEcosystem: 0.3,
  },
  elixir: {
    performanceTier: 3,
    concurrencyModel: 'actor-model',
    computeStrength: 0.4,
    ioStrength: 0.9,
    stringStrength: 0.7,
    mlEcosystem: 0.2,
    webEcosystem: 0.8,
    systemsEcosystem: 0.3,
  },
  haskell: {
    performanceTier: 3,
    concurrencyModel: 'STM-green-threads',
    computeStrength: 0.7,
    ioStrength: 0.5,
    stringStrength: 0.5,
    mlEcosystem: 0.2,
    webEcosystem: 0.3,
    systemsEcosystem: 0.4,
  },
  zig: {
    performanceTier: 1,
    concurrencyModel: 'manual',
    computeStrength: 1.0,
    ioStrength: 0.6,
    stringStrength: 0.3,
    mlEcosystem: 0.1,
    webEcosystem: 0.2,
    systemsEcosystem: 0.9,
  },
  scala: {
    performanceTier: 3,
    concurrencyModel: 'actors-futures',
    computeStrength: 0.6,
    ioStrength: 0.7,
    stringStrength: 0.6,
    mlEcosystem: 0.5,
    webEcosystem: 0.6,
    systemsEcosystem: 0.3,
  },
  ocaml: {
    performanceTier: 2,
    concurrencyModel: 'lwt-async',
    computeStrength: 0.8,
    ioStrength: 0.5,
    stringStrength: 0.5,
    mlEcosystem: 0.2,
    webEcosystem: 0.2,
    systemsEcosystem: 0.5,
  },
  php: {
    performanceTier: 4,
    concurrencyModel: 'request-per-process',
    computeStrength: 0.2,
    ioStrength: 0.6,
    stringStrength: 0.8,
    mlEcosystem: 0.1,
    webEcosystem: 0.8,
    systemsEcosystem: 0.0,
  },
};

/**
 * Analyze a function's characteristics from its topology.
 */
function analyzeFunctionCharacteristics(func: ExtractedFunction): {
  isComputeHeavy: boolean;
  isIOHeavy: boolean;
  isConcurrent: boolean;
  hasStringOps: boolean;
  complexity: 'simple' | 'moderate' | 'complex';
  domain: 'compute' | 'io' | 'web' | 'ml' | 'systems' | 'general';
} {
  const hasLoop = func.edgeTypes.includes('PROCESS') && func.nodeCount > 5;
  const hasFork = func.edgeTypes.includes('FORK');
  const hasRace = func.edgeTypes.includes('RACE');

  const isComputeHeavy = hasLoop && func.nodeCount > 8;
  const isIOHeavy = func.callees.some((c) =>
    /fetch|read|write|open|connect|request|http|sql|query|get|post/i.test(c)
  );
  const isConcurrent = hasFork || hasRace;
  const hasStringOps = func.callees.some((c) =>
    /format|parse|replace|split|join|regex|match|template/i.test(c)
  );

  const complexity =
    func.nodeCount <= 3
      ? 'simple'
      : func.nodeCount <= 10
      ? 'moderate'
      : 'complex';

  // Infer domain from function names and callees.
  const allNames = [func.name, ...func.callees].join(' ').toLowerCase();
  let domain: 'compute' | 'io' | 'web' | 'ml' | 'systems' | 'general' =
    'general';
  if (
    /train|predict|model|tensor|neural|gradient|embed|classify/.test(allNames)
  )
    domain = 'ml';
  else if (
    /http|api|route|handler|serve|request|response|middleware/.test(allNames)
  )
    domain = 'web';
  else if (
    /malloc|mmap|syscall|kernel|driver|socket|buffer|alloc/.test(allNames)
  )
    domain = 'systems';
  else if (isComputeHeavy) domain = 'compute';
  else if (isIOHeavy) domain = 'io';

  return {
    isComputeHeavy,
    isIOHeavy,
    isConcurrent,
    hasStringOps,
    complexity,
    domain,
  };
}

/**
 * Find the best language for a function based on its topology characteristics.
 *
 * Analyzes the function's control flow, callees, and complexity to determine
 * which language is the optimal fit.
 */
export function findBestLanguage(
  func: ExtractedFunction,
  candidateLanguages?: string[]
): LanguageFitness[] {
  const chars = analyzeFunctionCharacteristics(func);
  const candidates = candidateLanguages ?? Object.keys(LANGUAGE_PROFILES);

  const results: LanguageFitness[] = candidates.map((lang) => {
    const profile = LANGUAGE_PROFILES[lang];
    if (!profile) {
      return {
        language: lang,
        score: 0,
        breakdown: {
          computeFit: 0,
          ioFit: 0,
          concurrencyFit: 0,
          performanceTier: 5,
          ecosystemFit: 0,
        },
        rationale: `Unknown language: ${lang}`,
      };
    }

    // Score each dimension.
    const computeFit = chars.isComputeHeavy ? profile.computeStrength : 0.5;
    const ioFit = chars.isIOHeavy ? profile.ioStrength : 0.5;
    const concurrencyFit = chars.isConcurrent
      ? profile.concurrencyModel.includes('goroutine') ||
        profile.concurrencyModel.includes('actor') ||
        profile.concurrencyModel.includes('structured')
        ? 0.9
        : profile.concurrencyModel.includes('thread')
        ? 0.6
        : 0.4
      : 0.5;
    const ecosystemFit =
      chars.domain === 'ml'
        ? profile.mlEcosystem
        : chars.domain === 'web'
        ? profile.webEcosystem
        : chars.domain === 'systems'
        ? profile.systemsEcosystem
        : 0.5;

    // Performance weight.
    const perfScore = 1.0 - (profile.performanceTier - 1) / 4;

    // Weighted combination.
    const weights = {
      compute: chars.isComputeHeavy ? 0.35 : 0.15,
      io: chars.isIOHeavy ? 0.25 : 0.1,
      concurrency: chars.isConcurrent ? 0.2 : 0.05,
      performance: 0.2,
      ecosystem: 0.2,
    };

    const score =
      computeFit * weights.compute +
      ioFit * weights.io +
      concurrencyFit * weights.concurrency +
      perfScore * weights.performance +
      ecosystemFit * weights.ecosystem;

    // Build rationale.
    const strengths: string[] = [];
    if (computeFit > 0.7) strengths.push('strong compute');
    if (ioFit > 0.7) strengths.push('excellent IO');
    if (concurrencyFit > 0.7) strengths.push('native concurrency');
    if (ecosystemFit > 0.7) strengths.push(`${chars.domain} ecosystem`);
    if (perfScore > 0.7) strengths.push('high performance');

    const rationale =
      strengths.length > 0
        ? `${lang}: ${strengths.join(', ')}`
        : `${lang}: general purpose`;

    return {
      language: lang,
      score,
      breakdown: {
        computeFit,
        ioFit,
        concurrencyFit,
        performanceTier: profile.performanceTier,
        ecosystemFit,
      },
      rationale,
    };
  });

  return results.sort((a, b) => b.score - a.score);
}

/**
 * Find the least bule (minimum beta cost) expression of a topology.
 *
 * The beta cost of a topology is the sum of beta values across all edges:
 * - PROCESS: beta = 0 (sequential, no cost)
 * - FORK: beta = branch_count - 1 (concurrent branches)
 * - RACE: beta = alternatives - 1 (concurrent alternatives)
 * - FOLD: beta = sources - 1 (merge cost)
 * - VENT: beta = deficit (resource leak cost)
 *
 * The least bule expression minimizes total beta while preserving semantics.
 */
export function computeBetaCost(func: ExtractedFunction): {
  totalBeta: number;
  breakdown: Array<{ edgeType: string; beta: number; count: number }>;
  leastBuleScore: number;
} {
  const edgeCosts = new Map<string, { beta: number; count: number }>();

  for (const edge of func.ast.edges) {
    const existing = edgeCosts.get(edge.type) ?? { beta: 0, count: 0 };
    let edgeBeta = 0;

    switch (edge.type) {
      case 'PROCESS':
        edgeBeta = 0;
        break;
      case 'FORK':
        edgeBeta = Math.max(0, edge.targetIds.length - 1);
        break;
      case 'RACE':
        edgeBeta = Math.max(0, edge.targetIds.length - 1);
        break;
      case 'FOLD':
        edgeBeta = Math.max(0, edge.sourceIds.length - 1);
        break;
      case 'VENT':
        edgeBeta = 1; // Deficit = 1 per vent
        break;
      case 'SLIVER':
        edgeBeta = 1; // Lock contention = 1
        break;
      default:
        edgeBeta = 0;
    }

    existing.beta += edgeBeta;
    existing.count += 1;
    edgeCosts.set(edge.type, existing);
  }

  const totalBeta = [...edgeCosts.values()].reduce((sum, e) => sum + e.beta, 0);
  const breakdown = [...edgeCosts.entries()].map(([type, data]) => ({
    edgeType: type,
    beta: data.beta,
    count: data.count,
  }));

  // Least bule score: 0.0 = pure sequential (beta=0), 1.0 = maximally concurrent.
  // Normalized by node count.
  const maxPossibleBeta = Math.max(1, func.nodeCount - 1);
  const leastBuleScore = totalBeta / maxPossibleBeta;

  return { totalBeta, breakdown, leastBuleScore };
}

// ─── Topo-Race Language Search ───────────────────────────────────────────────

/**
 * Generate a topo-race topology: race the same function across N languages.
 *
 * Like x-gnosis's topo-race compression (race brotli vs gzip vs deflate),
 * but for programming languages. Run the same function in Python, Go, and Rust
 * simultaneously; the fastest correct result wins.
 */
export function generateTopoRaceTopology(
  functionName: string,
  _filePath: string,
  languages: string[]
): string {
  const lines: string[] = [];

  lines.push(`// Topo-race: ${functionName} across ${languages.join(', ')}`);
  lines.push(`// First correct result wins.`);
  lines.push('');
  lines.push(
    `(race_entry: Entry:PolyglotBridgeEntry { name='${functionName}_race' })`
  );
  lines.push('');

  for (const lang of languages) {
    const ext = languageExtension(lang);
    const nodeId = `${sanitizeId(functionName)}_${lang}`;
    lines.push(
      `(${nodeId}: Statement:PolyglotBridgeCall { ` +
        `language='${lang}', file='${toSnakeCase(
          functionName
        )}.${ext}', callee='${functionName}' })`
    );
  }
  lines.push('');

  lines.push(`(race_winner: Return:PolyglotBridgeReturn)`);
  lines.push('');

  // Entry RACE to all language implementations.
  const targets = languages
    .map((l) => `${sanitizeId(functionName)}_${l}`)
    .join(' | ');
  lines.push(`(race_entry)-[:RACE]->(${targets})`);

  // All implementations FOLD to winner.
  const sources = languages
    .map((l) => `${sanitizeId(functionName)}_${l}`)
    .join(' | ');
  lines.push(`(${sources})-[:FOLD]->(race_winner)`);

  lines.push('');
  return lines.join('\n');
}

// ─── Utilities ───────────────────────────────────────────────────────────────

function sanitizeId(name: string): string {
  return name.replace(/[^a-zA-Z0-9_]/g, '_');
}

function toSnakeCase(s: string): string {
  return s
    .replace(/([a-z])([A-Z])/g, '$1_$2')
    .replace(/[-\s]+/g, '_')
    .toLowerCase();
}

function languageExtension(language: string): string {
  const map: Record<string, string> = {
    python: 'py',
    javascript: 'mjs',
    typescript: 'ts',
    go: 'go',
    rust: 'rs',
    ruby: 'rb',
    lua: 'lua',
    php: 'php',
    elixir: 'exs',
    kotlin: 'kt',
    scala: 'scala',
    swift: 'swift',
    haskell: 'hs',
    ocaml: 'ml',
    c: 'c',
    cpp: 'cpp',
    c_sharp: 'cs',
    java: 'java',
    zig: 'zig',
    bash: 'sh',
  };
  return map[language] ?? 'txt';
}
