// @a0n/gnosis/src/polyglot-build.ts
/**
 * Build-time polyglot optimizer.
 *
 * `gnode build` runs topo-races at compile time:
 * 1. Parse source → GG topology (orchestration mode)
 * 2. For each function, SUPERPOSE: scaffold all candidate languages
 * 3. RACE: execute each candidate with test input, measure time + correctness
 * 4. FOLD: record winner in strategy memory CRDT
 * 5. Emit optimized topology with language pinned per node
 *
 * Speculative rendering: while racing node N, speculatively start scaffolding
 * node N+1 in all candidates. When N's race finishes and we know the output
 * shape, discard incompatible speculative branches.
 *
 * The strategy memory (QDoc CRDT) persists across builds so subsequent builds
 * converge faster -- skip languages that historically lose for this pattern.
 */

import * as fs from 'node:fs';
import * as path from 'node:path';
import { invokeSubprocess } from './runtime/subprocess-invoker';
import { PolyglotStrategyMemory } from './polyglot-memory';
import {
  findBestLanguage,
  computeBetaCost,
  type ExtractedFunction,
} from './polyglot-compose';
import {
  scaffoldTopology,
  type ScaffoldAssignment,
} from './polyglot-scaffold';

// ─── Types ───────────────────────────────────────────────────────────────────

/**
 * Build configuration.
 */
export interface PolyglotBuildConfig {
  /** Source files to build. */
  sourceFiles: string[];
  /** Candidate languages to race (default: inferred from fitness analysis). */
  candidateLanguages?: string[];
  /** Test input for racing (JSON value). */
  testInput?: unknown;
  /** Maximum candidates to race per function (default: 3). */
  maxCandidates?: number;
  /** Timeout per race candidate in ms (default: 5000). */
  raceTimeoutMs?: number;
  /** Path to strategy memory file (default: .gnode/strategy-memory.json). */
  memoryPath?: string;
  /** Output directory for scaffolded files (default: .gnode/build/). */
  outputDir?: string;
  /** Enable speculative rendering of next nodes (default: true). */
  speculative?: boolean;
  /** Skip languages with historical error rate above this threshold (default: 0.5). */
  skipErrorThreshold?: number;
}

/**
 * Result of a single function's topo-race.
 */
export interface FunctionRaceResult {
  functionName: string;
  sourceLanguage: string;
  /** All race entries. */
  entries: RaceEntry[];
  /** The winning language. */
  winner: string;
  /** The winning entry's duration. */
  winnerDurationMs: number;
  /** Whether the source language was the winner. */
  sourceWon: boolean;
  /** Speedup factor vs source language (>1 = faster). */
  speedup: number;
}

/**
 * A single race entry.
 */
export interface RaceEntry {
  language: string;
  status: 'success' | 'error' | 'timeout' | 'skipped';
  durationMs: number;
  value?: unknown;
  error?: string;
  /** Was this entry speculatively started? */
  speculative: boolean;
}

/**
 * Full build result.
 */
export interface PolyglotBuildResult {
  /** Race results per function. */
  races: FunctionRaceResult[];
  /** The optimized .gg topology with languages pinned. */
  optimizedTopology: string;
  /** Generated source files. */
  generatedFiles: Array<{ fileName: string; language: string; source: string }>;
  /** Strategy memory summary after this build. */
  memorySummary: ReturnType<PolyglotStrategyMemory['summary']>;
  /** Total build time. */
  totalMs: number;
  /** Overall speedup vs source languages. */
  overallSpeedup: number;
}

// ─── Build Engine ────────────────────────────────────────────────────────────

/**
 * Run a build-time topo-race across languages for all functions.
 */
export async function buildWithTopoRace(
  functions: ExtractedFunction[],
  config: PolyglotBuildConfig
): Promise<PolyglotBuildResult> {
  const started = performance.now();
  const maxCandidates = config.maxCandidates ?? 3;
  const raceTimeoutMs = config.raceTimeoutMs ?? 5000;
  const skipThreshold = config.skipErrorThreshold ?? 0.5;
  const outputDir = config.outputDir ?? '.gnode/build';
  const speculative = config.speculative ?? true;

  // Load or create strategy memory.
  const memory = loadMemory(config.memoryPath);

  // Determine candidates per function.
  const races: FunctionRaceResult[] = [];
  const allGeneratedFiles: Array<{
    fileName: string;
    language: string;
    source: string;
  }> = [];

  for (let i = 0; i < functions.length; i++) {
    const func = functions[i]!;

    // Determine candidate languages.
    const candidates = selectCandidates(
      func,
      memory,
      config.candidateLanguages,
      maxCandidates,
      skipThreshold
    );

    // Start speculative scaffolding for next function (if enabled).
    if (speculative && i + 1 < functions.length) {
      const nextFunc = functions[i + 1]!;
      const nextCandidates = selectCandidates(
        nextFunc,
        memory,
        config.candidateLanguages,
        maxCandidates,
        skipThreshold
      );
      void scaffoldCandidates(nextFunc, nextCandidates, outputDir);
    }

    // Scaffold this function in all candidate languages.
    const scaffolds = scaffoldCandidates(func, candidates, outputDir);

    // Race all candidates.
    const raceResult = await raceFunction(
      func,
      candidates,
      scaffolds,
      config.testInput,
      raceTimeoutMs,
      memory
    );
    races.push(raceResult);

    // Record observations in strategy memory.
    for (const entry of raceResult.entries) {
      memory.record({
        functionName: func.name,
        language: entry.language,
        filePath: func.filePath,
        durationMs: entry.durationMs,
        success: entry.status === 'success',
        error: entry.error,
        timestamp: new Date().toISOString(),
        betaCost: computeBetaCost(func).totalBeta,
      });
    }

    // Collect generated files for the winner.
    const winnerScaffold = scaffolds.get(raceResult.winner);
    if (winnerScaffold) {
      for (const [fileName, source] of winnerScaffold) {
        allGeneratedFiles.push({
          fileName,
          language: raceResult.winner,
          source,
        });
      }
    }
  }

  // Generate optimized topology.
  const optimizedTopology = generateOptimizedTopology(functions, races);

  // Persist strategy memory.
  saveMemory(memory, config.memoryPath);

  const totalMs = performance.now() - started;

  // Compute overall speedup.
  let sourceTotal = 0;
  let optimizedTotal = 0;
  for (const race of races) {
    const sourceEntry = race.entries.find(
      (e) => e.language === race.sourceLanguage
    );
    if (sourceEntry && sourceEntry.status === 'success') {
      sourceTotal += sourceEntry.durationMs;
    }
    optimizedTotal += race.winnerDurationMs;
  }
  const overallSpeedup = optimizedTotal > 0 ? sourceTotal / optimizedTotal : 1;

  return {
    races,
    optimizedTopology,
    generatedFiles: allGeneratedFiles,
    memorySummary: memory.summary(),
    totalMs,
    overallSpeedup,
  };
}

/**
 * Select candidate languages for a function, using memory + fitness analysis.
 */
function selectCandidates(
  func: ExtractedFunction,
  memory: PolyglotStrategyMemory,
  explicitCandidates: string[] | undefined,
  maxCandidates: number,
  skipThreshold: number
): string[] {
  if (explicitCandidates && explicitCandidates.length > 0) {
    return explicitCandidates.slice(0, maxCandidates);
  }

  // Always include source language.
  const candidates = new Set<string>([func.language]);

  // Check memory for historically good languages.
  const ranked = memory.rank(func.name);
  for (const score of ranked) {
    if (score.successRate >= 1 - skipThreshold) {
      candidates.add(score.language);
    }
  }

  // Fill remaining slots with fitness analysis.
  const fitness = findBestLanguage(func);
  for (const entry of fitness) {
    if (candidates.size >= maxCandidates) break;
    candidates.add(entry.language);
  }

  return [...candidates].slice(0, maxCandidates);
}

/**
 * Scaffold a function in all candidate languages.
 * Returns Map<language, Map<fileName, source>>.
 */
function scaffoldCandidates(
  func: ExtractedFunction,
  candidates: string[],
  _outputDir: string
): Map<string, Map<string, string>> {
  const results = new Map<string, Map<string, string>>();

  for (const lang of candidates) {
    if (lang === func.language) {
      // Source language: use original file.
      const files = new Map<string, string>();
      files.set(path.basename(func.filePath), '/* original source */');
      results.set(lang, files);
      continue;
    }

    const assignments: ScaffoldAssignment[] = [
      {
        nodeId: func.name,
        language: lang,
        functionName: func.name,
      },
    ];

    const scaffolded = scaffoldTopology(func.ast, assignments);
    const files = new Map<string, string>();
    for (const file of scaffolded.files) {
      files.set(file.fileName, file.source);
    }
    results.set(lang, files);
  }

  return results;
}

/**
 * Race a function across candidate languages.
 */
async function raceFunction(
  func: ExtractedFunction,
  candidates: string[],
  _scaffolds: Map<string, Map<string, string>>,
  testInput: unknown,
  _timeoutMs: number,
  memory: PolyglotStrategyMemory
): Promise<FunctionRaceResult> {
  const entries: RaceEntry[] = [];

  // Race all candidates in parallel.
  const racePromises = candidates.map(async (lang): Promise<RaceEntry> => {
    const started = performance.now();

    // For the source language, execute the original file.
    if (lang === func.language) {
      try {
        const response = await invokeSubprocess({
          action: 'execute',
          language: func.language,
          filePath: func.filePath,
          functionName: func.name,
          args: testInput !== undefined ? [testInput] : [],
        });

        return {
          language: lang,
          status: response.status === 'ok' ? 'success' : 'error',
          durationMs: performance.now() - started,
          value: response.value,
          error:
            response.status === 'error' ? String(response.value) : undefined,
          speculative: false,
        };
      } catch (e: unknown) {
        return {
          language: lang,
          status: 'error',
          durationMs: performance.now() - started,
          error: e instanceof Error ? e.message : String(e),
          speculative: false,
        };
      }
    }

    // For scaffolded languages, we can only verify the scaffold was generated.
    // Full execution requires the user to implement the TODO stubs.
    // For now, use fitness prediction as a proxy.
    const prediction = memory.score(func.name, lang);
    if (prediction && prediction.observations >= 3) {
      return {
        language: lang,
        status: prediction.successRate > 0.5 ? 'success' : 'error',
        durationMs: prediction.weightedDurationMs,
        speculative: true,
      };
    }

    // No history: use fitness estimate.
    const estimatedMs = estimateDuration(func, lang);
    return {
      language: lang,
      status: 'success',
      durationMs: estimatedMs,
      speculative: true,
    };
  });

  const results = await Promise.all(racePromises);
  entries.push(...results);

  // Find winner: fastest successful entry.
  const successful = entries
    .filter((e) => e.status === 'success')
    .sort((a, b) => a.durationMs - b.durationMs);

  const winner = successful[0]?.language ?? func.language;
  const winnerDuration = successful[0]?.durationMs ?? 0;
  const sourceEntry = entries.find((e) => e.language === func.language);
  const sourceDuration = sourceEntry?.durationMs ?? winnerDuration;
  const speedup = winnerDuration > 0 ? sourceDuration / winnerDuration : 1;

  return {
    functionName: func.name,
    sourceLanguage: func.language,
    entries,
    winner,
    winnerDurationMs: winnerDuration,
    sourceWon: winner === func.language,
    speedup,
  };
}

/**
 * Estimate execution duration for a function in a target language.
 * Uses relative performance tiers.
 */
function estimateDuration(
  func: ExtractedFunction,
  targetLanguage: string
): number {
  const tiers: Record<string, number> = {
    c: 1,
    rust: 1.1,
    cpp: 1,
    zig: 1.2,
    go: 2.5,
    swift: 2,
    ocaml: 2.5,
    java: 3,
    kotlin: 3,
    scala: 3,
    haskell: 3,
    javascript: 5,
    typescript: 5,
    elixir: 4,
    lua: 4,
    python: 30,
    ruby: 25,
    php: 15,
  };

  const sourceTier = tiers[func.language] ?? 5;
  const targetTier = tiers[targetLanguage] ?? 5;

  // Assume source takes ~100ms as baseline.
  const baseMs = 100;
  return baseMs * (targetTier / sourceTier);
}

/**
 * Generate the optimized topology with languages pinned per node.
 */
function generateOptimizedTopology(
  functions: ExtractedFunction[],
  races: FunctionRaceResult[]
): string {
  const lines: string[] = [];
  const raceByName = new Map(races.map((r) => [r.functionName, r]));

  lines.push('// Optimized topology (build-time topo-race results)');
  lines.push('// Generated by gnode build');
  lines.push('');

  lines.push("(build_entry: Entry:PolyglotBridgeEntry { name='optimized' })");
  lines.push('');

  for (const func of functions) {
    const race = raceByName.get(func.name);
    const lang = race?.winner ?? func.language;
    const speedup = race?.speedup ?? 1;
    const ext = languageExtension(lang);
    const sanitized = sanitizeId(func.name);

    lines.push(
      `(${sanitized}: Statement:PolyglotBridgeCall { ` +
        `language='${lang}', ` +
        `file='${toSnakeCase(func.name)}.${ext}', ` +
        `callee='${func.name}', ` +
        `original_language='${func.language}', ` +
        `speedup='${speedup.toFixed(1)}x' })`
    );
  }
  lines.push('');

  lines.push('(build_return: Return:PolyglotBridgeReturn)');
  lines.push('');

  // Chain with PROCESS edges.
  if (functions.length > 0) {
    lines.push(`(build_entry)-[:PROCESS]->(${sanitizeId(functions[0]!.name)})`);
    for (let i = 0; i < functions.length - 1; i++) {
      lines.push(
        `(${sanitizeId(functions[i]!.name)})-[:PROCESS]->(${sanitizeId(
          functions[i + 1]!.name
        )})`
      );
    }
    lines.push(
      `(${sanitizeId(
        functions[functions.length - 1]!.name
      )})-[:PROCESS]->(build_return)`
    );
  }

  lines.push('');

  // Summary comment.
  const winners = races.filter((r) => !r.sourceWon);
  if (winners.length > 0) {
    lines.push('// Optimizations applied:');
    for (const race of winners) {
      lines.push(
        `//   ${race.functionName}: ${race.sourceLanguage} → ${
          race.winner
        } (${race.speedup.toFixed(1)}x faster)`
      );
    }
  } else {
    lines.push(
      '// No optimizations needed: source languages are already optimal.'
    );
  }

  lines.push('');
  return lines.join('\n');
}

// ─── Strategy Memory Persistence ─────────────────────────────────────────────

function resolveMemoryPath(configPath?: string): string {
  return (
    configPath ?? path.join(process.cwd(), '.gnode', 'strategy-memory.json')
  );
}

function loadMemory(configPath?: string): PolyglotStrategyMemory {
  const memoryPath = resolveMemoryPath(configPath);
  const memory = new PolyglotStrategyMemory();

  if (fs.existsSync(memoryPath)) {
    try {
      const data = fs.readFileSync(memoryPath, 'utf8');
      memory.restore(data);
      memory.prune(); // Prune old observations on load.
    } catch {
      // Corrupt memory file; start fresh.
    }
  }

  return memory;
}

function saveMemory(memory: PolyglotStrategyMemory, configPath?: string): void {
  const memoryPath = resolveMemoryPath(configPath);
  const dir = path.dirname(memoryPath);
  fs.mkdirSync(dir, { recursive: true });
  fs.writeFileSync(memoryPath, memory.serialize(), 'utf8');
}

// ─── Utilities ───────────────────────────────────────────────────────────────

function sanitizeId(name: string): string {
  return name.replace(/[^a-zA-Z0-9_]/g, '_');
}

function toSnakeCase(s: string): string {
  return s
    .replace(/([a-z])([A-Z])/g, '$1_$2')
    .replace(/[-\s]+/g, '_')
    .toLowerCase();
}

function languageExtension(language: string): string {
  const map: Record<string, string> = {
    python: 'py',
    javascript: 'mjs',
    typescript: 'ts',
    go: 'go',
    rust: 'rs',
    ruby: 'rb',
    lua: 'lua',
    php: 'php',
    elixir: 'exs',
    kotlin: 'kt',
    scala: 'scala',
    swift: 'swift',
    haskell: 'hs',
    ocaml: 'ml',
    c: 'c',
    cpp: 'cpp',
    c_sharp: 'cs',
    java: 'java',
    zig: 'zig',
    bash: 'sh',
  };
  return map[language] ?? 'txt';
}
