// @a0n/gnosis/src/polyglot-scaffold.ts
/**
 * Topology-driven code scaffolder.
 *
 * Given a .gg topology with labeled nodes, generates skeleton implementations
 * in each language. The topology specifies the contract; the skeletons satisfy it.
 *
 * Usage:
 *   gnode scaffold pipeline.gg --python ml_predict --go api_handler --rust parser
 *
 * Generates:
 *   ml_predict.py     -- def ml_predict(input): ...
 *   api_handler.go    -- func apiHandler(input interface{}) interface{} { ... }
 *   parser.rs         -- pub fn parser(input: Value) -> Value { ... }
 */

import type { GraphAST } from './betty/compiler';

/**
 * Language-specific function signature and skeleton.
 */
export interface ScaffoldedFunction {
  language: string;
  fileName: string;
  functionName: string;
  /** Generated source code. */
  source: string;
  /** The node ID in the topology this function implements. */
  nodeId: string;
}

/**
 * Scaffold assignment: maps a node ID to a language.
 */
export interface ScaffoldAssignment {
  nodeId: string;
  language: string;
  functionName?: string;
}

/**
 * Result of scaffolding a topology.
 */
export interface ScaffoldResult {
  /** Generated files. */
  files: ScaffoldedFunction[];
  /** The original topology source with language annotations added. */
  annotatedTopology: string;
}

/**
 * File extension for each language.
 */
const LANGUAGE_EXTENSIONS: Record<string, string> = {
  python: 'py',
  javascript: 'mjs',
  typescript: 'ts',
  go: 'go',
  rust: 'rs',
  ruby: 'rb',
  lua: 'lua',
  php: 'php',
  elixir: 'exs',
  kotlin: 'kt',
  scala: 'scala',
  swift: 'swift',
  haskell: 'hs',
  ocaml: 'ml',
  c: 'c',
  cpp: 'cpp',
  c_sharp: 'cs',
  java: 'java',
  zig: 'zig',
  bash: 'sh',
};

export const POLYGLOT_SCAFFOLD_LANGUAGES = Object.freeze(
  Object.keys(LANGUAGE_EXTENSIONS).sort()
);

/**
 * Generate a function skeleton in the specified language.
 */
function generateSkeleton(
  language: string,
  functionName: string,
  nodeId: string,
  labels: string[],
  properties: Record<string, string>
): string {
  const description =
    properties.description ?? `Implements topology node '${nodeId}'`;

  switch (language) {
    case 'python':
      return `"""${description}"""


def ${functionName}(input_data):
    """
    Topology node: ${nodeId}
    Labels: ${labels.join(', ')}

    Args:
        input_data: Input from the previous topology node.

    Returns:
        Result to pass to the next topology node.
    """
    # TODO: implement
    return input_data
`;

    case 'go':
      return `package main

// ${description}
//
// Topology node: ${nodeId}
// Labels: ${labels.join(', ')}

import "encoding/json"

// ${toPascalCase(functionName)} implements the topology node.
func ${toPascalCase(functionName)}(input interface{}) interface{} {
\t// TODO: implement
\treturn input
}

func main() {
\t// Read JSON from stdin, call function, write JSON to stdout.
\tvar input interface{}
\tjson.NewDecoder(os.Stdin).Decode(&input)
\tresult := ${toPascalCase(functionName)}(input)
\tjson.NewEncoder(os.Stdout).Encode(result)
}
`;

    case 'rust':
      return `//! ${description}
//!
//! Topology node: ${nodeId}
//! Labels: ${labels.join(', ')}

use serde_json::Value;

/// Implements the topology node.
pub fn ${toSnakeCase(functionName)}(input: Value) -> Value {
    // TODO: implement
    input
}

fn main() {
    let input: Value = serde_json::from_reader(std::io::stdin()).unwrap();
    let result = ${toSnakeCase(functionName)}(input);
    serde_json::to_writer(std::io::stdout(), &result).unwrap();
}
`;

    case 'ruby':
      return `# ${description}
#
# Topology node: ${nodeId}
# Labels: ${labels.join(', ')}

def ${toSnakeCase(functionName)}(input_data)
  # TODO: implement
  input_data
end
`;

    case 'javascript':
    case 'typescript':
      return `/**
 * ${description}
 *
 * Topology node: ${nodeId}
 * Labels: ${labels.join(', ')}
 */

export function ${toCamelCase(functionName)}(inputData${
        language === 'typescript' ? ': unknown' : ''
      })${language === 'typescript' ? ': unknown' : ''} {
  // TODO: implement
  return inputData;
}
`;

    case 'java':
      return `/**
 * ${description}
 *
 * Topology node: ${nodeId}
 * Labels: ${labels.join(', ')}
 */
public class ${toPascalCase(functionName)} {
    public static Object ${toCamelCase(functionName)}(Object input) {
        // TODO: implement
        return input;
    }

    public static void main(String[] args) {
        // Read JSON from stdin, call function, write JSON to stdout.
    }
}
`;

    case 'c':
      return `/* ${description}
 *
 * Topology node: ${nodeId}
 * Labels: ${labels.join(', ')}
 */

#include <stdio.h>
#include <stdlib.h>

/* TODO: implement */
int main(void) {
    /* Read from stdin, process, write to stdout */
    return 0;
}
`;

    case 'lua':
      return `-- ${description}
--
-- Topology node: ${nodeId}
-- Labels: ${labels.join(', ')}

function ${toSnakeCase(functionName)}(input_data)
  -- TODO: implement
  return input_data
end
`;

    case 'swift':
      return `/// ${description}
///
/// Topology node: ${nodeId}
/// Labels: ${labels.join(', ')}

import Foundation

func ${toCamelCase(functionName)}(inputData: Any) -> Any {
    // TODO: implement
    return inputData
}
`;

    default:
      return `# ${description}\n# Topology node: ${nodeId}\n# Language: ${language}\n# TODO: implement\n`;
  }
}

/**
 * Scaffold skeleton implementations for a topology.
 *
 * @param ast The parsed GraphAST.
 * @param assignments Map of node IDs to languages.
 * @param ggSource Original .gg source (for annotation).
 */
export function scaffoldTopology(
  ast: GraphAST,
  assignments: ScaffoldAssignment[],
  ggSource?: string
): ScaffoldResult {
  const files: ScaffoldedFunction[] = [];

  for (const assignment of assignments) {
    const node = ast.nodes.get(assignment.nodeId);
    if (!node) {
      continue;
    }

    const functionName =
      assignment.functionName ??
      node.properties.callee ??
      node.properties.name ??
      assignment.nodeId;

    const ext = LANGUAGE_EXTENSIONS[assignment.language] ?? 'txt';
    const fileName = `${toSnakeCase(functionName)}.${ext}`;

    const source = generateSkeleton(
      assignment.language,
      functionName,
      assignment.nodeId,
      node.labels,
      node.properties
    );

    files.push({
      language: assignment.language,
      fileName,
      functionName,
      source,
      nodeId: assignment.nodeId,
    });
  }

  // Annotate the topology with language assignments.
  let annotatedTopology = ggSource ?? '';
  if (annotatedTopology) {
    for (const assignment of assignments) {
      const ext = LANGUAGE_EXTENSIONS[assignment.language] ?? 'txt';
      const funcName = assignment.functionName ?? assignment.nodeId;
      const fileName = `${toSnakeCase(funcName)}.${ext}`;
      // Add language/file properties to the node line.
      const nodePattern = new RegExp(
        `\\(${escapeRegex(assignment.nodeId)}:([^)]+)\\)`
      );
      const match = annotatedTopology.match(nodePattern);
      if (match) {
        const existingContent = match[1]!;
        if (existingContent.includes('{')) {
          // Already has properties -- add to them.
          annotatedTopology = annotatedTopology.replace(
            nodePattern,
            `(${assignment.nodeId}:${existingContent.replace(
              '}',
              `, language='${assignment.language}', file='${fileName}' }`
            )})`
          );
        } else {
          // No properties -- add them.
          annotatedTopology = annotatedTopology.replace(
            nodePattern,
            `(${assignment.nodeId}:${existingContent.trim()} { language='${
              assignment.language
            }', file='${fileName}' })`
          );
        }
      }
    }
  }

  return { files, annotatedTopology };
}

// ─── String Utilities ────────────────────────────────────────────────────────

function toCamelCase(s: string): string {
  return s.replace(/[-_](.)/g, (_, c: string) => c.toUpperCase());
}

function toPascalCase(s: string): string {
  const camel = toCamelCase(s);
  return camel.charAt(0).toUpperCase() + camel.slice(1);
}

function toSnakeCase(s: string): string {
  return s
    .replace(/([a-z])([A-Z])/g, '$1_$2')
    .replace(/[-\s]+/g, '_')
    .toLowerCase();
}

function escapeRegex(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}
