// @a0n/gnosis/src/polyglot-registry.ts
import { readFile } from 'node:fs/promises';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { POLYGLOT_SCAFFOLD_LANGUAGES } from './polyglot-scaffold';

export const POLYGLOT_OPERATIONS = [
  'analyze',
  'explain',
  'scaffold',
  'translate',
  'rewritePreview',
] as const;

export type PolyglotOperation = (typeof POLYGLOT_OPERATIONS)[number];
export type PolyglotCapabilityStatus =
  | 'supported'
  | 'experimental'
  | 'unsupported';

export interface PolyglotLanguageCapability {
  id: string;
  displayName: string;
  operations: Record<PolyglotOperation, PolyglotCapabilityStatus>;
}

export interface PolyglotCapabilityMatrix {
  registrySource: string;
  languages: PolyglotLanguageCapability[];
}

const REWRITE_PREVIEW_LANGUAGES = new Set([
  'typescript',
  'javascript',
  'python',
  'go',
  'rust',
  'java',
]);

function polyglotExtractorSourcePath(): string {
  return path.resolve(
    path.dirname(fileURLToPath(import.meta.url)),
    '..',
    'polyglot',
    'src',
    'extractors',
    'mod.rs'
  );
}

function formatLanguageDisplayName(id: string): string {
  return id
    .split('_')
    .map((segment) =>
      segment.length > 0
        ? segment.charAt(0).toUpperCase() + segment.slice(1)
        : segment
    )
    .join(' ');
}

function parseSupportedLanguagesFromRustSource(source: string): string[] {
  const match = source.match(
    /pub fn supported_languages\(\) -> Vec<&'static str>\s*\{\s*vec!\[(?<body>[\s\S]*?)\]\s*\}/m
  );
  if (!match?.groups?.body) {
    throw new Error(
      'Could not locate supported_languages() in polyglot extractor source'
    );
  }

  return Array.from(
    match.groups.body.matchAll(/"([^"]+)"/g),
    (entry) => {
      const language = entry[1];
      if (language === undefined) {
        throw new Error('Malformed supported_languages() entry');
      }
      return language;
    }
  ).filter((value, index, values) => values.indexOf(value) === index);
}

/**
 * Handles the gnosis get Polyglot Supported Languages workflow.
 */
export async function getPolyglotSupportedLanguages(): Promise<string[]> {
  const source = await readFile(polyglotExtractorSourcePath(), 'utf8');
  return parseSupportedLanguagesFromRustSource(source);
}

/**
 * Handles the gnosis get Polyglot Capability Matrix workflow.
 */
export async function getPolyglotCapabilityMatrix(): Promise<PolyglotCapabilityMatrix> {
  const supportedLanguages = await getPolyglotSupportedLanguages();
  const scaffoldLanguages = new Set(POLYGLOT_SCAFFOLD_LANGUAGES);

  return {
    registrySource:
      'open-source/gnosis/polyglot/src/extractors/mod.rs::supported_languages()',
    languages: supportedLanguages.map((id) => ({
      id,
      displayName: formatLanguageDisplayName(id),
      operations: {
        analyze: 'supported',
        explain: 'supported',
        scaffold: scaffoldLanguages.has(id) ? 'supported' : 'unsupported',
        translate: scaffoldLanguages.has(id) ? 'supported' : 'unsupported',
        rewritePreview: REWRITE_PREVIEW_LANGUAGES.has(id)
          ? 'experimental'
          : 'unsupported',
      },
    })),
  };
}

/**
 * Parses the Supported Languages For Test.
 */
export function parseSupportedLanguagesForTest(source: string): string[] {
  return parseSupportedLanguagesFromRustSource(source);
}
