// @a0n/gnosis/src/prismatic-refraction-duckdb.ts
/**
 * Portable SQL materialization of the Prismatic Refraction →
 * Knotgraph ingest paragon.
 *
 * Emits `CREATE TABLE` + parameterized `INSERT` statements + a bulk
 * `COPY`-equivalent in a SQL dialect that runs unchanged on:
 *
 *   - **DuckDB** — in-process analytics (@duckdb/node-api or
 *     @duckdb/duckdb-wasm in the browser)
 *   - **Cloudflare D1** — SQLite-backed edge DB
 *   - **SQLite / WebSQL-WASM** — browser-side via `sql.js` or
 *     `@sqlite.org/sqlite-wasm`
 *
 * The module is dependency-free: it produces SQL strings + typed
 * parameter arrays. The caller wires the connection. This keeps the
 * gnosis package ingest-neutral and lets the consumer pick a storage
 * lane without dragging in 50 MB of DuckDB binaries.
 *
 * The schema stores the 15 app-zone dim projections as `DOUBLE` columns
 * plus the full 54-dim `Float64Array` as a `BLOB`. The typed columns
 * are query-indexable; the blob is the bit-exact source of truth (so a
 * SELECT + deserialize round-trips to the original KnotgraphShapedBlock).
 *
 * Query affordances exposed as SQL string builders (not executed):
 *
 *   - `duckdbSelectByTenant(tenantHi, tenantLo)`  — tenant-isolated scan
 *   - `duckdbSelectByRegion(tenant, regionId)`    — lane-scoped scan
 *   - `duckdbSelectByEntity(tenant, entityId)`    — content-addressed get
 */

import type {
  IngestResult,
  KnotgraphShapedBlock,
  TenantId,
} from './prismatic-refraction-knotgraph';
import { KNOTGRAPH_DIMS } from './prismatic-refraction-knotgraph';

// ─── Schema ──────────────────────────────────────────────────────────

export const DUCKDB_TABLE_NAME = 'knotgraph_blocks' as const;

/** Portable CREATE TABLE. `DOUBLE` is accepted by DuckDB (fp64 exact)
 *  and by SQLite/D1 (stored as `REAL`, 8-byte IEEE-754). `BLOB` is the
 *  bit-exact full `Float64Array`. */
export function duckdbSchemaDDL(): string {
  return `CREATE TABLE IF NOT EXISTS ${DUCKDB_TABLE_NAME} (
  tenant_hi       DOUBLE NOT NULL,
  tenant_lo       DOUBLE NOT NULL,
  region_id       DOUBLE NOT NULL,
  type_tag        DOUBLE NOT NULL,
  entity_id       DOUBLE NOT NULL,
  schema_knot     DOUBLE NOT NULL,
  parent_commit   DOUBLE NOT NULL,
  field_0         DOUBLE NOT NULL,
  field_1         DOUBLE NOT NULL,
  field_2         DOUBLE NOT NULL,
  field_3         DOUBLE NOT NULL,
  field_4         DOUBLE NOT NULL,
  field_5         DOUBLE NOT NULL,
  provenance_did  DOUBLE NOT NULL,
  freshness_clock DOUBLE NOT NULL,
  dims_blob       BLOB NOT NULL
);`;
}

/** Tenant-indexed + region-indexed lookup indexes. Idempotent. */
export function duckdbIndexDDL(): string[] {
  return [
    `CREATE INDEX IF NOT EXISTS idx_${DUCKDB_TABLE_NAME}_tenant ON ${DUCKDB_TABLE_NAME} (tenant_hi, tenant_lo);`,
    `CREATE INDEX IF NOT EXISTS idx_${DUCKDB_TABLE_NAME}_region ON ${DUCKDB_TABLE_NAME} (tenant_hi, tenant_lo, region_id);`,
    `CREATE INDEX IF NOT EXISTS idx_${DUCKDB_TABLE_NAME}_entity ON ${DUCKDB_TABLE_NAME} (tenant_hi, tenant_lo, entity_id);`,
  ];
}

// ─── Parameter row ───────────────────────────────────────────────────

/** Ordered parameter tuple for the 16-slot INSERT. Column order matches
 *  `duckdbInsertSQL()`. Typed so consumers can't permute accidentally. */
export interface BlockRow {
  readonly tenantHi: number;
  readonly tenantLo: number;
  readonly regionId: number;
  readonly typeTag: number;
  readonly entityId: number;
  readonly schemaKnot: number;
  readonly parentCommit: number;
  readonly field0: number;
  readonly field1: number;
  readonly field2: number;
  readonly field3: number;
  readonly field4: number;
  readonly field5: number;
  readonly provenanceDid: number;
  readonly freshnessClock: number;
  readonly dimsBlob: Uint8Array;
}

/** Project a KnotgraphShapedBlock into a DB row. The `dimsBlob` is the
 *  underlying Float64Array's byte view (8 × 54 = 432 bytes) — bit-exact. */
export function blockToRow(block: KnotgraphShapedBlock): BlockRow {
  const d = block.dims;
  const D = KNOTGRAPH_DIMS;
  return {
    tenantHi: d[D.TENANT_ID_HI]!,
    tenantLo: d[D.TENANT_ID_LO]!,
    regionId: d[D.REGION_ID]!,
    typeTag: d[D.TYPE_TAG]!,
    entityId: d[D.ENTITY_ID]!,
    schemaKnot: d[D.SCHEMA_KNOT_HASH]!,
    parentCommit: d[D.PARENT_COMMIT]!,
    field0: d[D.FIELD_0]!,
    field1: d[D.FIELD_1]!,
    field2: d[D.FIELD_2]!,
    field3: d[D.FIELD_3]!,
    field4: d[D.FIELD_4]!,
    field5: d[D.FIELD_5]!,
    provenanceDid: d[D.PROVENANCE_DID]!,
    freshnessClock: d[D.FRESHNESS_CLOCK]!,
    dimsBlob: new Uint8Array(d.buffer, d.byteOffset, d.byteLength),
  };
}

/** Rebuild a KnotgraphShapedBlock from a stored `dimsBlob`. Round-trip
 *  inverse of `blockToRow`. */
export function rowBlobToBlock(dimsBlob: Uint8Array): KnotgraphShapedBlock {
  if (dimsBlob.byteLength !== 54 * 8) {
    throw new RangeError(
      `dimsBlob must be 432 bytes (54 × 8), got ${dimsBlob.byteLength}`,
    );
  }
  // Copy into a detached buffer so callers can't mutate shared state.
  const copy = new Uint8Array(dimsBlob.byteLength);
  copy.set(dimsBlob);
  const dims = new Float64Array(copy.buffer);
  return { dims };
}

// ─── Parameterized INSERT ────────────────────────────────────────────

export const DUCKDB_INSERT_PLACEHOLDERS = 16 as const;

/** The INSERT template. 16 positional placeholders — DuckDB uses `?`,
 *  D1 uses `?`, sqlite-wasm uses `?`; same string works everywhere. */
export function duckdbInsertSQL(): string {
  return `INSERT INTO ${DUCKDB_TABLE_NAME} (
  tenant_hi, tenant_lo, region_id, type_tag, entity_id,
  schema_knot, parent_commit,
  field_0, field_1, field_2, field_3, field_4, field_5,
  provenance_did, freshness_clock, dims_blob
) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?);`;
}

/** Positional parameter array for a single INSERT. Order matches
 *  `duckdbInsertSQL()`. Sixteen elements. */
export function rowToInsertParams(row: BlockRow): ReadonlyArray<number | Uint8Array> {
  return [
    row.tenantHi,
    row.tenantLo,
    row.regionId,
    row.typeTag,
    row.entityId,
    row.schemaKnot,
    row.parentCommit,
    row.field0,
    row.field1,
    row.field2,
    row.field3,
    row.field4,
    row.field5,
    row.provenanceDid,
    row.freshnessClock,
    row.dimsBlob,
  ];
}

/** Bulk materialization: one prepared INSERT executed N times. Returns
 *  SQL + a params array per block. Consumers execute with their
 *  connection's prepared-statement API:
 *
 *   - DuckDB:   `const stmt = conn.prepare(sql); for (p of params) stmt.run(p);`
 *   - D1:       `const ps = env.D1.prepare(sql); await env.D1.batch(params.map(p => ps.bind(...p)))`
 *   - SQLite:   `const stmt = db.prepare(sql); for (p of params) stmt.run(p);`
 */
export function materializeIngest(
  result: IngestResult,
): {
  ddl: string;
  indexes: string[];
  insertSql: string;
  params: ReadonlyArray<ReadonlyArray<number | Uint8Array>>;
} {
  const rows = result.blocks.map(blockToRow);
  return {
    ddl: duckdbSchemaDDL(),
    indexes: duckdbIndexDDL(),
    insertSql: duckdbInsertSQL(),
    params: rows.map(rowToInsertParams),
  };
}

// ─── Query affordances (string builders, not executors) ──────────────

/** Tenant-isolated full scan. Returns (sql, params). Always filters by
 *  tenant to preserve the knotgraph `TenantIsolation` invariant at the
 *  SQL layer. */
export function duckdbSelectByTenant(
  tenant: TenantId,
): { sql: string; params: ReadonlyArray<number> } {
  return {
    sql: `SELECT tenant_hi, tenant_lo, region_id, type_tag, entity_id,
       schema_knot, parent_commit,
       field_0, field_1, field_2, field_3, field_4, field_5,
       provenance_did, freshness_clock, dims_blob
FROM ${DUCKDB_TABLE_NAME}
WHERE tenant_hi = ? AND tenant_lo = ?;`,
    params: [tenant.hi, tenant.lo],
  };
}

/** Lane-scoped scan: all blocks for (tenant, region). The region maps
 *  to the prismatic-refraction laneId, so this is effectively
 *  "get all blocks on wire lane R." */
export function duckdbSelectByRegion(
  tenant: TenantId,
  regionId: number,
): { sql: string; params: ReadonlyArray<number> } {
  return {
    sql: `SELECT tenant_hi, tenant_lo, region_id, type_tag, entity_id,
       schema_knot, parent_commit,
       field_0, field_1, field_2, field_3, field_4, field_5,
       provenance_did, freshness_clock, dims_blob
FROM ${DUCKDB_TABLE_NAME}
WHERE tenant_hi = ? AND tenant_lo = ? AND region_id = ?;`,
    params: [tenant.hi, tenant.lo, regionId],
  };
}

/** Content-addressed get: (tenant, entity_id) uniquely identifies the
 *  latest version of an entity. Expect ≤ 1 row. */
export function duckdbSelectByEntity(
  tenant: TenantId,
  entityId: number,
): { sql: string; params: ReadonlyArray<number> } {
  return {
    sql: `SELECT tenant_hi, tenant_lo, region_id, type_tag, entity_id,
       schema_knot, parent_commit,
       field_0, field_1, field_2, field_3, field_4, field_5,
       provenance_did, freshness_clock, dims_blob
FROM ${DUCKDB_TABLE_NAME}
WHERE tenant_hi = ? AND tenant_lo = ? AND entity_id = ?;`,
    params: [tenant.hi, tenant.lo, entityId],
  };
}

// ─── Alternate storage lanes (doc-only, no code emitted) ─────────────

/**
 * For Cloudflare D1:
 *
 *   const ps = env.D1.prepare(insertSql);
 *   await env.D1.batch(params.map((p) => ps.bind(...p)));
 *
 * D1 accepts the same SQL unchanged (it is SQLite). D1's
 * `prepare().bind().all()` is the read path for the SELECT builders.
 *
 * For WebSQL / sqlite-wasm (browser):
 *
 *   import sqlite3InitModule from '@sqlite.org/sqlite-wasm';
 *   const sqlite3 = await sqlite3InitModule();
 *   const db = new sqlite3.oo1.DB();
 *   db.exec(ddl);
 *   for (const stmt of indexes) db.exec(stmt);
 *   const prepared = db.prepare(insertSql);
 *   for (const p of params) { prepared.bind(p); prepared.step(); prepared.reset(); }
 *
 * The SQL dialect is identical. The only difference is the connection
 * API surface.
 */
