// @a0n/gnosis/src/prismatic-refraction.ts
/**
 * Prismatic Refraction — TypeScript mirror of
 * `ForkRaceFoldTheorems.PrismaticRefraction` (zero sorry, `import Init` only).
 *
 * Canonical SSOT for the "vomit bits, delegate shape to geometry" wire
 * protocol:
 *
 *   - `RawStream` — app-side byte firehose (no shape obligation)
 *   - `prism` — wire-layer content-addressed lane assignment
 *   - `refract` — lane-compatible pairwise fold on encounter
 *   - `betaOneDeficit` — outstanding-shard count (proxy for unclosed loops)
 *   - `StorageGate` — reader-side admissibility filter (knotgraph
 *     tenant-isolation analog)
 *
 * Each exported function carries a doc comment referencing the Lean
 * theorem it embodies so runtime consumers can cite the formal contract.
 * The Lean source is Nat-valued; this module uses `number` under
 * non-negative-integer discipline.
 */

// ─── Part I: Raw Byte Stream (App Contract) ──────────────────────────

/** The app-side emitter output: a pure byte firehose. No shape obligation.
 *  Mirrors `PrismaticRefraction.RawStream`. */
export interface RawStream {
  readonly bytes: ReadonlyArray<number>;
}

/** Construct a RawStream. No structural validation — the app layer is
 *  shapeless by contract (`THM-APP-CONTRACT-IS-SHAPELESS`). Individual
 *  byte values are expected to be `0..255` but are not range-checked;
 *  the lane-assignment is well-defined for any non-negative integer. */
export function createRawStream(bytes: ReadonlyArray<number>): RawStream {
  return { bytes };
}

// ─── Part II: Wire-Layer Prism (Content-Addressed Lane Assignment) ───

/** A shard is a lane-tagged, phase-tagged byte-payload atom.
 *  Mirrors `PrismaticRefraction.Shard`. */
export interface Shard {
  readonly laneId: number;
  readonly phase: number;
  readonly payload: ReadonlyArray<number>;
}

/** Content-addressed lane assignment: lane is `byte % laneCount`.
 *  Mirrors `assignLane`. The runtime may substitute XXH64 without
 *  violating the theorems — the formal module only requires determinism
 *  and the lane bound, both of which `mod` provides. */
export function assignLane(byte: number, laneCount: number): number {
  if (laneCount <= 0) {
    throw new RangeError(`laneCount must be positive, got ${laneCount}`);
  }
  return ((byte % laneCount) + laneCount) % laneCount;
}

/** Build a single-byte micro-shard. Mirrors `byteToShard`. */
export function byteToShard(byte: number, laneCount: number): Shard {
  return {
    laneId: assignLane(byte, laneCount),
    phase: 0,
    payload: [byte],
  };
}

/** The wire-layer prism: map each byte to its lane-tagged micro-shard.
 *  Mirrors `prism` and is covered by:
 *    - `THM-PRISM-IS-DETERMINISTIC`
 *    - `THM-PRISM-LANE-COUNT`     — output length = input length
 *    - `THM-PRISM-PRESERVES-BYTE-COUNT` — total payload = input length
 *    - `THM-PRISM-RESPECTS-LANE-BOUND`  — every laneId ∈ [0, laneCount)
 */
export function prism(stream: RawStream, laneCount: number): Shard[] {
  if (laneCount <= 0) {
    throw new RangeError(`laneCount must be positive, got ${laneCount}`);
  }
  const out: Shard[] = new Array(stream.bytes.length);
  for (let i = 0; i < stream.bytes.length; i++) {
    out[i] = byteToShard(stream.bytes[i]!, laneCount);
  }
  return out;
}

/** Sum of payload lengths across a shard list. Mirrors `totalPayloadLength`. */
export function totalPayloadLength(shards: ReadonlyArray<Shard>): number {
  let total = 0;
  for (const s of shards) total += s.payload.length;
  return total;
}

// ─── Part III: Encounter Refraction (Lane-Compatible Fold) ───────────

/** Compatibility predicate: same lane ⇒ refract succeeds.
 *  Mirrors `foldCompatible`. */
export function foldCompatible(a: Shard, b: Shard): boolean {
  return a.laneId === b.laneId;
}

/** Refraction: fold two compatible shards into one by concatenating
 *  payloads and summing phases. Returns `null` when lanes differ.
 *  Mirrors `refract`, covered by:
 *    - `THM-REFRACT-PRESERVES-PAYLOAD-ON-FOLD`
 *    - `THM-REFRACT-RETURNS-NONE-ON-INCOMPATIBLE`
 *    - `THM-REFRACT-COMMUTES-ON-SINGLE-LANE` (up to phase + length) */
export function refract(a: Shard, b: Shard): Shard | null {
  if (a.laneId !== b.laneId) return null;
  return {
    laneId: a.laneId,
    phase: a.phase + b.phase,
    payload: [...a.payload, ...b.payload],
  };
}

// ─── Part IV: Disjoint-Lane Commutativity ────────────────────────────

/** Two refractions on disjoint lanes produce lane-disjoint results.
 *  Mirrors `THM-DISJOINT-LANE-ENCOUNTERS-COMMUTE`. Returns `null` when
 *  either precondition fails; otherwise returns both results. */
export function disjointLaneFold(
  a: Shard,
  b: Shard,
  c: Shard,
  d: Shard,
): { r1: Shard; r2: Shard } | null {
  if (a.laneId !== b.laneId) return null;
  if (c.laneId !== d.laneId) return null;
  if (a.laneId === c.laneId) return null;
  const r1 = refract(a, b);
  const r2 = refract(c, d);
  if (r1 === null || r2 === null) return null;
  return { r1, r2 };
}

// ─── Part V: β₁ Deficit ──────────────────────────────────────────────

/** β₁ deficit: outstanding-shard count. Proxy for unclosed loops.
 *  Mirrors `betaOneDeficit`. */
export function betaOneDeficit(shards: ReadonlyArray<Shard>): number {
  return shards.length;
}

/** Attempt a head-pair refraction on a shard list. On success, replaces
 *  the first two with their fold (β₁ deficit -1); on failure, leaves
 *  the list unchanged (deficit unchanged). Mirrors the operational
 *  content of `THM-SUCCESSFUL-REFRACTION-DECREASES-DEFICIT` and
 *  `THM-REFRACTION-NEVER-INCREASES-DEFICIT`. */
export function refractStep(shards: ReadonlyArray<Shard>): {
  result: Shard[];
  folded: boolean;
} {
  if (shards.length < 2) return { result: [...shards], folded: false };
  const r = refract(shards[0]!, shards[1]!);
  if (r === null) return { result: [...shards], folded: false };
  return { result: [r, ...shards.slice(2)], folded: true };
}

/** Greedily reduce a shard list by rotating to the next compatible pair
 *  and folding. Terminates when no compatible pair remains. The output
 *  is a canonical per-lane aggregation: one shard per distinct lane,
 *  payload is the concatenation of all byte-order-preserving inputs on
 *  that lane. This is the settled shape knotgraph ingests.
 *
 *  Observationally equivalent to iterating `refract` pairwise within
 *  each lane; uses mutable accumulators internally to stay O(n). The
 *  formal module reasons about the pairwise `refract`; this function is
 *  the operational shortcut with identical output semantics. */
export function settleByLane(shards: ReadonlyArray<Shard>): Shard[] {
  const payloads = new Map<number, number[]>();
  const phases = new Map<number, number>();
  for (const s of shards) {
    const buf = payloads.get(s.laneId);
    if (buf === undefined) {
      payloads.set(s.laneId, [...s.payload]);
      phases.set(s.laneId, s.phase);
    } else {
      for (const b of s.payload) buf.push(b);
      phases.set(s.laneId, (phases.get(s.laneId) ?? 0) + s.phase);
    }
  }
  return [...payloads.keys()]
    .sort((x, y) => x - y)
    .map((laneId) => ({
      laneId,
      phase: phases.get(laneId) ?? 0,
      payload: payloads.get(laneId)!,
    }));
}

// ─── Part VI: Gauge-Invisibility Completion Certificate ──────────────

/** A shard is gauge-invisible relative to a set of others if every other
 *  shard is on a different lane — no refraction with any of them will
 *  succeed. Mirrors `THM-GAUGE-INVISIBLE-IS-COMPLETE`: invisibility IS
 *  the completion certificate. */
export function isGaugeInvisible(
  s: Shard,
  others: ReadonlyArray<Shard>,
): boolean {
  for (const o of others) {
    if (o.laneId === s.laneId) return false;
  }
  return true;
}

// ─── Part VII: Storage Gate (Knotgraph-Compatible) ───────────────────

/** The storage gate: tenant ID + admissible-lane whitelist. Mirrors the
 *  knotgraph `TenantIsolation` pattern — tenant IS the admissibility
 *  gate. Corresponds to `PrismaticRefraction.StorageGate`. */
export interface StorageGate {
  readonly admissibleLanes: ReadonlyArray<number>;
  readonly tenantId: number;
}

/** Storage assembly: filter shards by admissible lanes. The reader-side
 *  prism-inverse. Mirrors `storageAssembly` and is covered by:
 *    - `THM-STORAGE-ONLY-SEES-ADMISSIBLE-LANES`
 *    - `THM-STORAGE-ASSEMBLY-IS-IDEMPOTENT` (membership form) */
export function storageAssembly(
  gate: StorageGate,
  shards: ReadonlyArray<Shard>,
): Shard[] {
  const admissible = new Set(gate.admissibleLanes);
  return shards.filter((s) => admissible.has(s.laneId));
}

/** Construct a StorageGate with basic integrity checks. */
export function createStorageGate(
  admissibleLanes: ReadonlyArray<number>,
  tenantId: number,
): StorageGate {
  for (const lane of admissibleLanes) {
    if (!Number.isInteger(lane) || lane < 0) {
      throw new RangeError(`admissibleLanes must be non-negative integers, got ${lane}`);
    }
  }
  if (!Number.isInteger(tenantId) || tenantId < 0) {
    throw new RangeError(`tenantId must be a non-negative integer, got ${tenantId}`);
  }
  return { admissibleLanes, tenantId };
}

// ─── Part VIII: Storage-Shape Soundness (Crown) ──────────────────────

/** Witness record returned by `wireToStorage`. Every field corresponds
 *  to a conjunct of `THM-WIRE-TO-STORAGE-IS-SOUND`. */
export interface WireToStorageWitness {
  /** Conjunct 1: every shard's lane < laneCount. */
  readonly lanesBounded: boolean;
  /** Conjunct 2: total payload = input byte count. */
  readonly payloadPreserved: boolean;
  /** Payload total in bytes. */
  readonly totalBytes: number;
  /** Shard count (pre-settle). */
  readonly shardCount: number;
  /** Post-settle shard list (one per distinct lane). */
  readonly settled: Shard[];
}

/** End-to-end wire pipeline: raw stream → prism → settle. Returns the
 *  settled shards plus a witness of the soundness invariants. Mirrors
 *  `THM-WIRE-TO-STORAGE-IS-SOUND`. The `laneCount ≤ 16` check enforces
 *  the knotgraph app-zone width (dims 38-53). */
export function wireToStorage(
  stream: RawStream,
  laneCount: number,
): WireToStorageWitness {
  if (laneCount <= 0) {
    throw new RangeError(`laneCount must be positive, got ${laneCount}`);
  }
  if (laneCount > 16) {
    throw new RangeError(
      `laneCount must be ≤ 16 to fit the knotgraph app zone (dims 38-53), got ${laneCount}`,
    );
  }
  const shards = prism(stream, laneCount);
  const totalBytes = totalPayloadLength(shards);
  let lanesBounded = true;
  for (const s of shards) {
    if (s.laneId < 0 || s.laneId >= laneCount) {
      lanesBounded = false;
      break;
    }
  }
  return {
    lanesBounded,
    payloadPreserved: totalBytes === stream.bytes.length,
    totalBytes,
    shardCount: shards.length,
    settled: settleByLane(shards),
  };
}
