// @a0n/gnosis/src/probability-routing.ts
import {
  selectProbabilityRoutes,
  type ProbabilityRoutePolicy,
  type ProbabilityRouteSelection,
  type ProbabilityRouteWitness,
  type RouteMassInput,
} from '@a0n/aeon-logic/probability-routing';
import {
  createProbabilityProcessCertificate,
  createProbabilityTraceCertificate,
  observeProbabilityRuntime,
  type ProbabilityRuntimeObserverDecision,
  type ProbabilityTraceCertificate,
} from './probability-certificate';

export {
  selectProbabilityRoutes,
  type ProbabilityRoutePolicy,
  type ProbabilityRouteSelection,
  type ProbabilityRouteWitness,
  type RouteMassInput,
};

export interface RuntimeProbabilityRouteInput {
  readonly certificateId: string;
  readonly target: string;
  readonly routes: readonly RouteMassInput[];
  readonly policy?: ProbabilityRoutePolicy;
}

export interface RuntimeProbabilityRouteDecision {
  readonly selection: ProbabilityRouteSelection;
  readonly trace: ProbabilityTraceCertificate;
  readonly observer: ProbabilityRuntimeObserverDecision;
}

/**
 * Creates the Runtime Probability Route Decision.
 */
export function createRuntimeProbabilityRouteDecision(
  input: RuntimeProbabilityRouteInput
): RuntimeProbabilityRouteDecision {
  const selection = selectProbabilityRoutes(input.routes, input.policy);
  const auditedWitnesses =
    selection.selected.length > 0 ? selection.selected : selection.witnesses;
  const steps = auditedWitnesses.map((witness) =>
    createProbabilityProcessCertificate({
      certificateId: `${input.certificateId}:${witness.route}`,
      target: `${input.target}:${witness.route}`,
      inputMass: witness.totalMass,
      outputMass: witness.count,
      massLoss: witness.residualMass,
      residual: {
        unobservedMass: witness.decision === 'cover' ? witness.residualMass : 0,
        truncatedMass: 0,
        coarseningDebt: witness.decision === 'cover' ? 0 : witness.residualMass,
      },
      residualBudget: input.policy?.residualBudget ?? witness.totalMass,
    })
  );
  const trace = createProbabilityTraceCertificate({
    certificateId: input.certificateId,
    target: input.target,
    steps,
    mode: 'additive',
    residualBudget: input.policy?.residualBudget ?? selection.totalMass,
  });

  return {
    selection,
    trace,
    observer: observeProbabilityRuntime(trace, {
      residualBudget: input.policy?.residualBudget ?? selection.totalMass,
    }),
  };
}
