// @a0n/gnosis/src/probability-certificate.ts
/**
 * Finite probability runtime certificates.
 *
 * Runtime mirror of `Gnosis.FiniteProbabilityCore.RuntimeCertificate`.
 * The surface deliberately uses safe natural-number integers only: no floating
 * confidence scores, no hidden normalization, and no silent mass loss.
 */

export const PROBABILITY_CERTIFICATE_VERSION =
  'gnosis-finite-probability-certificate/v1' as const;

export const PROBABILITY_CERTIFICATE_THEOREM_REFS = [
  'Gnosis.FiniteProbabilityCore.RuntimeProbabilityProcessCertificate',
  'Gnosis.FiniteProbabilityCore.runtime_process_certificate_no_hidden_defect',
  'Gnosis.FiniteProbabilityCore.RuntimeProbabilityKernelCertificate',
  'Gnosis.FiniteProbabilityCore.RuntimeProbabilityCoverCertificate',
  'Gnosis.FiniteProbabilityCore.runtime_cover_certificate_no_hidden_defect',
  'Gnosis.FiniteProbabilityCore.RuntimeCompletedInterfaceCertificate',
  'Gnosis.FiniteProbabilityCore.runtime_completed_interface_certificate_no_hidden_defect',
] as const;

export type ProbabilityCertificateVersion =
  typeof PROBABILITY_CERTIFICATE_VERSION;

export type ProbabilityCertificateKind =
  | 'process'
  | 'kernel'
  | 'cover'
  | 'completed-interface'
  | 'trace';

export type ProbabilityResidualBadge =
  | 'lossless'
  | 'bounded-residual'
  | 'needs-cover'
  | 'human-escalation';

export type ProbabilityRuntimeDecisionKind =
  | 'continue'
  | 'cover'
  | 'shrink-horizon'
  | 'human-escalation';

export interface ProbabilityResidual {
  readonly unobservedMass: number;
  readonly truncatedMass: number;
  readonly coarseningDebt: number;
}

export interface ProbabilityCertificateBase {
  readonly version: ProbabilityCertificateVersion;
  readonly certificateId: string;
  readonly target: string;
  readonly kind: ProbabilityCertificateKind;
  readonly theoremRefs: readonly string[];
}

export interface ProbabilityProcessCertificate
  extends ProbabilityCertificateBase {
  readonly kind: 'process';
  readonly inputMass: number;
  readonly outputMass: number;
  readonly massLoss: number;
  readonly residual: ProbabilityResidual;
  readonly residualTotal: number;
  readonly balanceHolds: boolean;
  readonly lossCovered: boolean;
  readonly badge: ProbabilityResidualBadge;
}

export interface ProbabilityKernelCertificate extends ProbabilityCertificateBase {
  readonly kind: 'kernel';
  readonly inputMass: number;
  readonly outputMass: number;
  readonly lostMass: number;
  readonly balanceHolds: boolean;
  readonly outputMassWithinInput: boolean;
}

export interface ProbabilityCoverCertificate extends ProbabilityCertificateBase {
  readonly kind: 'cover';
  readonly footprint: number;
  readonly shadow: number;
  readonly horizonDepth: number;
  readonly horizonResidualBudget: number;
  readonly footprintBounded: boolean;
  readonly shadowBounded: boolean;
  readonly badge: ProbabilityResidualBadge;
}

export interface ProbabilityCompletedInterfaceCertificate
  extends ProbabilityCertificateBase {
  readonly kind: 'completed-interface';
  readonly visibleDepth: number;
  readonly shadow: number;
  readonly horizonDepth: number;
  readonly horizonResidualBudget: number;
  readonly depthWithinHorizon: boolean;
  readonly shadowWithinHorizon: boolean;
  readonly badge: ProbabilityResidualBadge;
}

export interface ProbabilityTraceCertificate extends ProbabilityCertificateBase {
  readonly kind: 'trace';
  readonly steps: readonly ProbabilityProcessCertificate[];
  readonly inputMass: number;
  readonly outputMass: number;
  readonly massLoss: number;
  readonly residual: ProbabilityResidual;
  readonly residualTotal: number;
  readonly balanceHolds: boolean;
  readonly lossCovered: boolean;
  readonly badge: ProbabilityResidualBadge;
}

export type ProbabilityCertificate =
  | ProbabilityProcessCertificate
  | ProbabilityKernelCertificate
  | ProbabilityCoverCertificate
  | ProbabilityCompletedInterfaceCertificate
  | ProbabilityTraceCertificate;

export interface ProbabilityRuntimeObserverOptions {
  readonly residualBudget?: number;
  readonly escalationBudget?: number;
}

export interface ProbabilityRuntimeObserverDecision {
  readonly decision: ProbabilityRuntimeDecisionKind;
  readonly certificateId: string;
  readonly target: string;
  readonly residualTotal: number;
  readonly residualBudget: number | null;
  readonly escalationBudget: number | null;
  readonly reason: string;
}

export interface CreateProbabilityProcessCertificateInput {
  readonly certificateId: string;
  readonly target: string;
  readonly inputMass: number;
  readonly outputMass: number;
  readonly massLoss: number;
  readonly residual?: Partial<ProbabilityResidual>;
  readonly residualBudget?: number;
  readonly escalationBudget?: number;
  readonly theoremRefs?: readonly string[];
}

export interface CreateProbabilityKernelCertificateInput {
  readonly certificateId: string;
  readonly target: string;
  readonly inputMass: number;
  readonly outputMass: number;
  readonly lostMass: number;
  readonly theoremRefs?: readonly string[];
}

export interface CreateProbabilityCoverCertificateInput {
  readonly certificateId: string;
  readonly target: string;
  readonly footprint: number;
  readonly shadow: number;
  readonly horizonDepth: number;
  readonly horizonResidualBudget: number;
  readonly escalationBudget?: number;
  readonly theoremRefs?: readonly string[];
}

export interface CreateProbabilityCompletedInterfaceCertificateInput {
  readonly certificateId: string;
  readonly target: string;
  readonly visibleDepth: number;
  readonly shadow: number;
  readonly horizonDepth: number;
  readonly horizonResidualBudget: number;
  readonly escalationBudget?: number;
  readonly theoremRefs?: readonly string[];
}

export interface CreateProbabilityTraceCertificateInput {
  readonly certificateId: string;
  readonly target: string;
  readonly steps: readonly ProbabilityProcessCertificate[];
  readonly mode?: 'sequential' | 'additive';
  readonly residualBudget?: number;
  readonly escalationBudget?: number;
  readonly theoremRefs?: readonly string[];
}

function assertNonEmptyString(value: string, fieldName: string): void {
  if (value.trim().length === 0) {
    throw new Error(`${fieldName} must be a non-empty string`);
  }
}

function assertNat(value: number, fieldName: string): void {
  if (!Number.isSafeInteger(value) || value < 0) {
    throw new Error(`${fieldName} must be a safe natural-number integer`);
  }
}

function addNat(left: number, right: number, fieldName: string): number {
  const sum = left + right;
  assertNat(sum, fieldName);
  return sum;
}

function normalizeResidual(
  residual: Partial<ProbabilityResidual> | undefined,
): ProbabilityResidual {
  const normalized = {
    unobservedMass: residual?.unobservedMass ?? 0,
    truncatedMass: residual?.truncatedMass ?? 0,
    coarseningDebt: residual?.coarseningDebt ?? 0,
  };
  assertNat(normalized.unobservedMass, 'residual.unobservedMass');
  assertNat(normalized.truncatedMass, 'residual.truncatedMass');
  assertNat(normalized.coarseningDebt, 'residual.coarseningDebt');
  return normalized;
}

/**
 * Handles the gnosis probability Residual Total workflow.
 */
export function probabilityResidualTotal(
  residual: ProbabilityResidual,
): number {
  return addNat(
    addNat(residual.unobservedMass, residual.truncatedMass, 'residualSubtotal'),
    residual.coarseningDebt,
    'residualTotal',
  );
}

function theoremRefsWithDefaults(
  theoremRefs: readonly string[] | undefined,
): readonly string[] {
  return theoremRefs ?? PROBABILITY_CERTIFICATE_THEOREM_REFS;
}

function badgeForResidual(
  residualTotal: number,
  residualBudget: number | undefined,
  escalationBudget: number | undefined,
  covered: boolean,
): ProbabilityResidualBadge {
  if (residualTotal === 0 && covered) {
    return 'lossless';
  }
  if (!covered) {
    return 'needs-cover';
  }
  if (escalationBudget !== undefined && residualTotal > escalationBudget) {
    return 'human-escalation';
  }
  if (residualBudget !== undefined && residualTotal > residualBudget) {
    return 'needs-cover';
  }
  return 'bounded-residual';
}

/**
 * Creates the Probability Process Certificate.
 */
export function createProbabilityProcessCertificate(
  input: CreateProbabilityProcessCertificateInput,
): ProbabilityProcessCertificate {
  assertNonEmptyString(input.certificateId, 'certificateId');
  assertNonEmptyString(input.target, 'target');
  assertNat(input.inputMass, 'inputMass');
  assertNat(input.outputMass, 'outputMass');
  assertNat(input.massLoss, 'massLoss');
  if (input.residualBudget !== undefined) {
    assertNat(input.residualBudget, 'residualBudget');
  }
  if (input.escalationBudget !== undefined) {
    assertNat(input.escalationBudget, 'escalationBudget');
  }

  const residual = normalizeResidual(input.residual);
  const residualTotal = probabilityResidualTotal(residual);
  const balanceHolds = input.outputMass + input.massLoss === input.inputMass;
  const lossCovered = input.massLoss <= residualTotal;

  return {
    version: PROBABILITY_CERTIFICATE_VERSION,
    certificateId: input.certificateId,
    target: input.target,
    kind: 'process',
    inputMass: input.inputMass,
    outputMass: input.outputMass,
    massLoss: input.massLoss,
    residual,
    residualTotal,
    balanceHolds,
    lossCovered,
    badge: badgeForResidual(
      residualTotal,
      input.residualBudget,
      input.escalationBudget,
      balanceHolds && lossCovered,
    ),
    theoremRefs: theoremRefsWithDefaults(input.theoremRefs),
  };
}

/**
 * Handles the gnosis assert Probability Process Certificate workflow.
 */
export function assertProbabilityProcessCertificate(
  certificate: ProbabilityProcessCertificate,
): void {
  if (!certificate.balanceHolds) {
    throw new Error(
      `${certificate.certificateId} violates outputMass + massLoss = inputMass`,
    );
  }
  if (!certificate.lossCovered) {
    throw new Error(
      `${certificate.certificateId} violates massLoss <= residualTotal`,
    );
  }
}

/**
 * Creates the Probability Kernel Certificate.
 */
export function createProbabilityKernelCertificate(
  input: CreateProbabilityKernelCertificateInput,
): ProbabilityKernelCertificate {
  assertNonEmptyString(input.certificateId, 'certificateId');
  assertNonEmptyString(input.target, 'target');
  assertNat(input.inputMass, 'inputMass');
  assertNat(input.outputMass, 'outputMass');
  assertNat(input.lostMass, 'lostMass');
  const balanceHolds = input.outputMass + input.lostMass === input.inputMass;

  return {
    version: PROBABILITY_CERTIFICATE_VERSION,
    certificateId: input.certificateId,
    target: input.target,
    kind: 'kernel',
    inputMass: input.inputMass,
    outputMass: input.outputMass,
    lostMass: input.lostMass,
    balanceHolds,
    outputMassWithinInput: input.outputMass <= input.inputMass,
    theoremRefs: theoremRefsWithDefaults(input.theoremRefs),
  };
}

/**
 * Creates the Probability Cover Certificate.
 */
export function createProbabilityCoverCertificate(
  input: CreateProbabilityCoverCertificateInput,
): ProbabilityCoverCertificate {
  assertNonEmptyString(input.certificateId, 'certificateId');
  assertNonEmptyString(input.target, 'target');
  assertNat(input.footprint, 'footprint');
  assertNat(input.shadow, 'shadow');
  assertNat(input.horizonDepth, 'horizonDepth');
  assertNat(input.horizonResidualBudget, 'horizonResidualBudget');
  if (input.escalationBudget !== undefined) {
    assertNat(input.escalationBudget, 'escalationBudget');
  }
  const footprintBounded = input.footprint <= input.horizonDepth;
  const shadowBounded = input.shadow <= input.horizonResidualBudget;

  return {
    version: PROBABILITY_CERTIFICATE_VERSION,
    certificateId: input.certificateId,
    target: input.target,
    kind: 'cover',
    footprint: input.footprint,
    shadow: input.shadow,
    horizonDepth: input.horizonDepth,
    horizonResidualBudget: input.horizonResidualBudget,
    footprintBounded,
    shadowBounded,
    badge: badgeForResidual(
      input.shadow,
      input.horizonResidualBudget,
      input.escalationBudget,
      footprintBounded && shadowBounded,
    ),
    theoremRefs: theoremRefsWithDefaults(input.theoremRefs),
  };
}

/**
 * Creates the Probability Completed Interface Certificate.
 */
export function createProbabilityCompletedInterfaceCertificate(
  input: CreateProbabilityCompletedInterfaceCertificateInput,
): ProbabilityCompletedInterfaceCertificate {
  assertNonEmptyString(input.certificateId, 'certificateId');
  assertNonEmptyString(input.target, 'target');
  assertNat(input.visibleDepth, 'visibleDepth');
  assertNat(input.shadow, 'shadow');
  assertNat(input.horizonDepth, 'horizonDepth');
  assertNat(input.horizonResidualBudget, 'horizonResidualBudget');
  if (input.escalationBudget !== undefined) {
    assertNat(input.escalationBudget, 'escalationBudget');
  }
  const depthWithinHorizon = input.visibleDepth <= input.horizonDepth;
  const shadowWithinHorizon = input.shadow <= input.horizonResidualBudget;

  return {
    version: PROBABILITY_CERTIFICATE_VERSION,
    certificateId: input.certificateId,
    target: input.target,
    kind: 'completed-interface',
    visibleDepth: input.visibleDepth,
    shadow: input.shadow,
    horizonDepth: input.horizonDepth,
    horizonResidualBudget: input.horizonResidualBudget,
    depthWithinHorizon,
    shadowWithinHorizon,
    badge: badgeForResidual(
      input.shadow,
      input.horizonResidualBudget,
      input.escalationBudget,
      depthWithinHorizon && shadowWithinHorizon,
    ),
    theoremRefs: theoremRefsWithDefaults(input.theoremRefs),
  };
}

/**
 * Creates the Probability Trace Certificate.
 */
export function createProbabilityTraceCertificate(
  input: CreateProbabilityTraceCertificateInput,
): ProbabilityTraceCertificate {
  assertNonEmptyString(input.certificateId, 'certificateId');
  assertNonEmptyString(input.target, 'target');
  if (input.steps.length === 0) {
    throw new Error('steps must contain at least one process certificate');
  }
  if (input.residualBudget !== undefined) {
    assertNat(input.residualBudget, 'residualBudget');
  }
  if (input.escalationBudget !== undefined) {
    assertNat(input.escalationBudget, 'escalationBudget');
  }

  for (const step of input.steps) {
    assertProbabilityProcessCertificate(step);
  }

  const mode = input.mode ?? 'sequential';
  const first = input.steps[0]!;
  const last = input.steps[input.steps.length - 1]!;
  let inputMass = mode === 'sequential' ? first.inputMass : 0;
  let outputMass = mode === 'sequential' ? last.outputMass : 0;
  let massLoss = 0;
  let residual: ProbabilityResidual = {
    unobservedMass: 0,
    truncatedMass: 0,
    coarseningDebt: 0,
  };

  for (const step of input.steps) {
    if (mode === 'additive') {
      inputMass = addNat(inputMass, step.inputMass, 'trace.inputMass');
      outputMass = addNat(outputMass, step.outputMass, 'trace.outputMass');
    }
    massLoss = addNat(massLoss, step.massLoss, 'trace.massLoss');
    residual = {
      unobservedMass: addNat(
        residual.unobservedMass,
        step.residual.unobservedMass,
        'trace.residual.unobservedMass',
      ),
      truncatedMass: addNat(
        residual.truncatedMass,
        step.residual.truncatedMass,
        'trace.residual.truncatedMass',
      ),
      coarseningDebt: addNat(
        residual.coarseningDebt,
        step.residual.coarseningDebt,
        'trace.residual.coarseningDebt',
      ),
    };
  }

  const residualTotal = probabilityResidualTotal(residual);
  const balanceHolds = outputMass + massLoss === inputMass;
  const lossCovered = massLoss <= residualTotal;

  return {
    version: PROBABILITY_CERTIFICATE_VERSION,
    certificateId: input.certificateId,
    target: input.target,
    kind: 'trace',
    steps: input.steps,
    inputMass,
    outputMass,
    massLoss,
    residual,
    residualTotal,
    balanceHolds,
    lossCovered,
    badge: badgeForResidual(
      residualTotal,
      input.residualBudget,
      input.escalationBudget,
      balanceHolds && lossCovered,
    ),
    theoremRefs: theoremRefsWithDefaults(input.theoremRefs),
  };
}

function certificateResidualTotal(certificate: ProbabilityCertificate): number {
  switch (certificate.kind) {
    case 'process':
    case 'trace':
      return certificate.residualTotal;
    case 'cover':
      return certificate.shadow;
    case 'completed-interface':
      return certificate.shadow;
    case 'kernel':
      return certificate.lostMass;
  }
}

function certificateBadge(
  certificate: ProbabilityCertificate,
): ProbabilityResidualBadge {
  switch (certificate.kind) {
    case 'process':
    case 'cover':
    case 'completed-interface':
    case 'trace':
      return certificate.badge;
    case 'kernel':
      return certificate.balanceHolds && certificate.outputMassWithinInput
        ? certificate.lostMass === 0
          ? 'lossless'
          : 'bounded-residual'
        : 'needs-cover';
  }
}

/**
 * Handles the gnosis observe Probability Runtime workflow.
 */
export function observeProbabilityRuntime(
  certificate: ProbabilityCertificate,
  options: ProbabilityRuntimeObserverOptions = {},
): ProbabilityRuntimeObserverDecision {
  if (options.residualBudget !== undefined) {
    assertNat(options.residualBudget, 'residualBudget');
  }
  if (options.escalationBudget !== undefined) {
    assertNat(options.escalationBudget, 'escalationBudget');
  }

  const residualTotal = certificateResidualTotal(certificate);
  const residualBudget = options.residualBudget ?? null;
  const escalationBudget = options.escalationBudget ?? null;
  const badge = certificateBadge(certificate);

  if (
    badge === 'human-escalation' ||
    (escalationBudget !== null && residualTotal > escalationBudget)
  ) {
    return {
      decision: 'human-escalation',
      certificateId: certificate.certificateId,
      target: certificate.target,
      residualTotal,
      residualBudget,
      escalationBudget,
      reason: 'Residual mass exceeds the configured escalation budget.',
    };
  }

  if (
    badge === 'needs-cover' ||
    (residualBudget !== null && residualTotal > residualBudget)
  ) {
    return {
      decision: 'cover',
      certificateId: certificate.certificateId,
      target: certificate.target,
      residualTotal,
      residualBudget,
      escalationBudget,
      reason: 'Residual mass is not covered by the current finite horizon.',
    };
  }

  if (
    residualTotal > 0 &&
    (certificate.kind === 'cover' ||
      certificate.kind === 'completed-interface' ||
      certificate.kind === 'trace')
  ) {
    return {
      decision: 'shrink-horizon',
      certificateId: certificate.certificateId,
      target: certificate.target,
      residualTotal,
      residualBudget,
      escalationBudget,
      reason: 'Residual mass is bounded and can be reduced by narrowing the visible horizon.',
    };
  }

  return {
    decision: 'continue',
    certificateId: certificate.certificateId,
    target: certificate.target,
    residualTotal,
    residualBudget,
    escalationBudget,
    reason: 'Probability residual is lossless or already bounded for this runtime step.',
  };
}
