// @a0n/gnosis/src/protocol69.ts
import {
  type Stage,
  type CascadeReport,
  stage,
  overallRatio,
  compose,
  homomorphismHolds,
  cascadeReport,
  lawHolds,
  summarize,
} from './compressor-cascade';
import {
  IOTA_TREE_LEAN_MODULE,
  graphAstToIotaTree,
  iotaTreeApplicationCount,
  iotaTreeLeafCount,
  iotaTreeToGg,
  type IotaTree,
} from './iota-trees';
import { BettyCompiler } from './betty/compiler';

export {
  type Stage,
  type CascadeReport,
  stage,
  overallRatio,
  compose,
  homomorphismHolds,
  cascadeReport,
  lawHolds,
  summarize,
};

export const PROTOCOL69_SCHEMA_VERSION = 'gnosis.protocol69.v1';
export const PROTOCOL69_NAME = 'protocol69';
export const PROTOCOL69_WORD_NAME = 'sixty9';
export const PROTOCOL69_OPERATOR = 'xor';
export const MONSTER_NUMBER_SYMBOL_MAGIC = 'GNOSIS-MONSTER-NUMBER/1';

export const PROTOCOL69_BROADCAST_SYMBOL = 66;
export const PROTOCOL69_LOCAL_FOIL_OPERAND = 7;
export const PROTOCOL69_EXPECTED_PROJECTION = 69;

// OSI Layer classification: Protocol69 is L5 Session, L5AuthUnit.credExchange
export const PROTOCOL69_OSI_LAYER = 5;
export type PROTOCOL69_OSI_SUBSYSTEM = 'authentication';
export const PROTOCOL69_OSI_SUBSYSTEM: PROTOCOL69_OSI_SUBSYSTEM = 'authentication';
export type PROTOCOL69_OSI_UNIT = 'credExchange';
export const PROTOCOL69_OSI_UNIT: PROTOCOL69_OSI_UNIT = 'credExchange';

export type Protocol69Name = typeof PROTOCOL69_NAME;
export type Protocol69WordName = typeof PROTOCOL69_WORD_NAME;
export type Protocol69Operator = typeof PROTOCOL69_OPERATOR;
export type Protocol69SymbolMode = 'monster-number';

export interface Protocol69ProjectionEnvelope {
  readonly schemaVersion: typeof PROTOCOL69_SCHEMA_VERSION;
  readonly protocol: Protocol69Name;
  readonly wordName: Protocol69WordName;
  readonly mode: Protocol69SymbolMode;
  readonly magic: typeof MONSTER_NUMBER_SYMBOL_MAGIC;
  readonly broadcastSymbol: typeof PROTOCOL69_BROADCAST_SYMBOL;
  readonly localFoilOperand: typeof PROTOCOL69_LOCAL_FOIL_OPERAND;
  readonly operator: Protocol69Operator;
  readonly expectedProjection: typeof PROTOCOL69_EXPECTED_PROJECTION;
  readonly ascii: string;
  // OSI Layer classification
  readonly osiLayer: typeof PROTOCOL69_OSI_LAYER;
  readonly osiSubsystem: PROTOCOL69_OSI_SUBSYSTEM;
  readonly osiUnit?: PROTOCOL69_OSI_UNIT;
}

export const PROTOCOL69_IOTA_PROGRAM_SCHEMA_VERSION =
  'gnosis.protocol69.iota-program-teleport.v1';
export const PROTOCOL69_IOTA_PROGRAM_RECEIPT_SCHEMA_VERSION =
  'gnosis.protocol69.iota-program-cache-receipt.v1';
export const PROTOCOL69_IOTA_PROGRAM_CACHE_KEY_PREFIX =
  'gnosis:protocol69:iota-program-cache:v1';

export interface Protocol69InRoomRfSession {
  readonly transport: 'in-room-rf';
  readonly radio: 'pluto-sdr';
  readonly antennaCount: number;
  readonly roomId?: string;
  readonly carrierHz?: number;
  readonly attenuationDb?: number;
}

export interface Protocol69IotaProgramKeyMaterial {
  readonly leanModule: typeof IOTA_TREE_LEAN_MODULE;
  readonly sourceNodeCount: number;
  readonly ggNodeCount: number;
  readonly ggEdgeCount: number;
  readonly projectionOperand: typeof PROTOCOL69_EXPECTED_PROJECTION;
  readonly localOperand: typeof PROTOCOL69_LOCAL_FOIL_OPERAND;
  readonly programChecksum: string;
}

export interface Protocol69IotaProgramTeleportEnvelope {
  readonly schemaVersion: typeof PROTOCOL69_IOTA_PROGRAM_SCHEMA_VERSION;
  readonly teleportKind: 'iota-program';
  readonly session: Protocol69InRoomRfSession;
  readonly projection: Protocol69ProjectionEnvelope;
  readonly leanModule: typeof IOTA_TREE_LEAN_MODULE;
  readonly ggSource: string;
  readonly programChecksum: string;
  readonly keyMaterial: Protocol69IotaProgramKeyMaterial;
  readonly computeCacheKey: string;
  readonly leafCount: number;
  readonly applicationCount: number;
  readonly nodeCount: number;
  readonly projectionOperand: typeof PROTOCOL69_EXPECTED_PROJECTION;
}

export interface Protocol69IotaProgramCacheReceipt {
  readonly schemaVersion: typeof PROTOCOL69_IOTA_PROGRAM_RECEIPT_SCHEMA_VERSION;
  readonly teleportKind: 'iota-program-cache-key';
  readonly session: Protocol69InRoomRfSession;
  readonly projection: Protocol69ProjectionEnvelope;
  readonly leanModule: typeof IOTA_TREE_LEAN_MODULE;
  readonly programChecksum: string;
  readonly keyMaterial: Protocol69IotaProgramKeyMaterial;
  readonly computeCacheKey: string;
  readonly ggNodeCount: number;
  readonly ggEdgeCount: number;
  readonly leafCount: number;
  readonly applicationCount: number;
  readonly nodeCount: number;
}

/**
 * Handles the gnosis protocol69 Projection workflow.
 */
export function protocol69Projection(
  broadcastSymbol = PROTOCOL69_BROADCAST_SYMBOL,
  localFoilOperand = PROTOCOL69_LOCAL_FOIL_OPERAND,
): number {
  return (broadcastSymbol ^ localFoilOperand) >>> 0;
}

/**
 * Builds the Protocol69 Projection Envelope.
 */
export function buildProtocol69ProjectionEnvelope(): Protocol69ProjectionEnvelope {
  const expectedProjection = protocol69Projection(
    PROTOCOL69_BROADCAST_SYMBOL,
    PROTOCOL69_LOCAL_FOIL_OPERAND,
  );
  if (expectedProjection !== PROTOCOL69_EXPECTED_PROJECTION) {
    throw new Error(
      `protocol69 projection drift: expected ${PROTOCOL69_EXPECTED_PROJECTION}, got ${expectedProjection}`,
    );
  }
  return {
    schemaVersion: PROTOCOL69_SCHEMA_VERSION,
    protocol: PROTOCOL69_NAME,
    wordName: PROTOCOL69_WORD_NAME,
    mode: 'monster-number',
    magic: MONSTER_NUMBER_SYMBOL_MAGIC,
    broadcastSymbol: PROTOCOL69_BROADCAST_SYMBOL,
    localFoilOperand: PROTOCOL69_LOCAL_FOIL_OPERAND,
    operator: PROTOCOL69_OPERATOR,
    expectedProjection: PROTOCOL69_EXPECTED_PROJECTION,
    ascii: [
      MONSTER_NUMBER_SYMBOL_MAGIC,
      `schemaVersion=${PROTOCOL69_SCHEMA_VERSION}`,
      `protocol=${PROTOCOL69_NAME}`,
      `wordName=${PROTOCOL69_WORD_NAME}`,
      `broadcastSymbol=${PROTOCOL69_BROADCAST_SYMBOL}`,
      `localFoilOperand=${PROTOCOL69_LOCAL_FOIL_OPERAND}`,
      `operator=${PROTOCOL69_OPERATOR}`,
      `expectedProjection=${PROTOCOL69_EXPECTED_PROJECTION}`,
    ].join(' '),
    osiLayer: PROTOCOL69_OSI_LAYER,
    osiSubsystem: PROTOCOL69_OSI_SUBSYSTEM,
    osiUnit: PROTOCOL69_OSI_UNIT,
  };
}

/**
 * Handles the gnosis protocol69 Stable Checksum workflow.
 */
export function protocol69StableChecksum(source: string): string {
  let hash = 0x811c9dc5;
  for (let index = 0; index < source.length; index += 1) {
    hash ^= source.charCodeAt(index);
    hash = Math.imul(hash, 0x01000193) >>> 0;
  }
  return hash.toString(16).padStart(8, '0');
}

/**
 * Handles the gnosis protocol69 Iota Program Key Material workflow.
 */
export function protocol69IotaProgramKeyMaterial(
  programChecksum: string,
  leafCount: number,
  applicationCount: number
): Protocol69IotaProgramKeyMaterial {
  const sourceNodeCount = leafCount + applicationCount;
  return {
    leanModule: IOTA_TREE_LEAN_MODULE,
    sourceNodeCount,
    ggNodeCount: sourceNodeCount + 1,
    ggEdgeCount: applicationCount + applicationCount + 1,
    projectionOperand: PROTOCOL69_EXPECTED_PROJECTION,
    localOperand: PROTOCOL69_LOCAL_FOIL_OPERAND,
    programChecksum,
  };
}

function protocol69IotaStableKeyMaterial(
  keyMaterial: Protocol69IotaProgramKeyMaterial
): string {
  return [
    keyMaterial.leanModule,
    keyMaterial.sourceNodeCount,
    keyMaterial.ggNodeCount,
    keyMaterial.ggEdgeCount,
    keyMaterial.projectionOperand,
    keyMaterial.localOperand,
    keyMaterial.programChecksum,
  ].join('|');
}

/**
 * Handles the gnosis protocol69 Iota Program Compute Cache Key workflow.
 */
export function protocol69IotaProgramComputeCacheKey(
  ggSource: string,
  session: Protocol69InRoomRfSession = protocol69DefaultInRoomRfSession(),
  keyMaterial?: Protocol69IotaProgramKeyMaterial
): string {
  const material =
    keyMaterial ??
    protocol69IotaProgramKeyMaterial(protocol69StableChecksum(ggSource), 1, 0);
  const sessionMaterial = [
    session.transport,
    session.radio,
    session.antennaCount,
    session.roomId ?? '',
    session.carrierHz ?? '',
    session.attenuationDb ?? '',
  ].join('|');
  return `${PROTOCOL69_IOTA_PROGRAM_CACHE_KEY_PREFIX}:${protocol69StableChecksum(
    `${ggSource}\0${protocol69IotaStableKeyMaterial(material)}\0${sessionMaterial}`
  )}`;
}

/**
 * Handles the gnosis protocol69 Default In Room Rf Session workflow.
 */
export function protocol69DefaultInRoomRfSession(): Protocol69InRoomRfSession {
  return {
    transport: 'in-room-rf',
    radio: 'pluto-sdr',
    antennaCount: 2,
  };
}

/**
 * Builds the Protocol69 Iota Program Teleport Envelope.
 */
export function buildProtocol69IotaProgramTeleportEnvelope(
  program: IotaTree,
  session: Protocol69InRoomRfSession = protocol69DefaultInRoomRfSession()
): Protocol69IotaProgramTeleportEnvelope {
  const projection = buildProtocol69ProjectionEnvelope();
  const ggSource = iotaTreeToGg(program);
  const leafCount = iotaTreeLeafCount(program);
  const applicationCount = iotaTreeApplicationCount(program);
  const programChecksum = protocol69StableChecksum(ggSource);
  const keyMaterial = protocol69IotaProgramKeyMaterial(
    programChecksum,
    leafCount,
    applicationCount
  );
  return {
    schemaVersion: PROTOCOL69_IOTA_PROGRAM_SCHEMA_VERSION,
    teleportKind: 'iota-program',
    session,
    projection,
    leanModule: IOTA_TREE_LEAN_MODULE,
    ggSource,
    programChecksum,
    keyMaterial,
    computeCacheKey: protocol69IotaProgramComputeCacheKey(
      ggSource,
      session,
      keyMaterial
    ),
    leafCount,
    applicationCount,
    nodeCount: leafCount + applicationCount,
    projectionOperand: PROTOCOL69_EXPECTED_PROJECTION,
  };
}

/**
 * Handles the gnosis validate Protocol69 Iota Program Teleport Envelope workflow.
 */
export function validateProtocol69IotaProgramTeleportEnvelope(
  envelope: Protocol69IotaProgramTeleportEnvelope
): Protocol69IotaProgramTeleportEnvelope {
  if (envelope.schemaVersion !== PROTOCOL69_IOTA_PROGRAM_SCHEMA_VERSION) {
    throw new Error('iota program teleport schemaVersion mismatch');
  }
  if (envelope.teleportKind !== 'iota-program') {
    throw new Error('iota program teleport kind mismatch');
  }
  validateProtocol69InRoomRfSession(envelope.session);
  if (envelope.leanModule !== IOTA_TREE_LEAN_MODULE) {
    throw new Error('iota program Lean anchor mismatch');
  }
  if (envelope.projectionOperand !== PROTOCOL69_EXPECTED_PROJECTION) {
    throw new Error('iota program projection operand mismatch');
  }
  if (
    envelope.projection.expectedProjection !== PROTOCOL69_EXPECTED_PROJECTION ||
    protocol69Projection(
      envelope.projection.broadcastSymbol,
      envelope.projection.localFoilOperand
    ) !== envelope.projection.expectedProjection
  ) {
    throw new Error('iota program projection envelope mismatch');
  }
  if (protocol69StableChecksum(envelope.ggSource) !== envelope.programChecksum) {
    throw new Error('iota program checksum mismatch');
  }
  const expectedKeyMaterial = protocol69IotaProgramKeyMaterial(
    envelope.programChecksum,
    envelope.leafCount,
    envelope.applicationCount
  );
  if (
    envelope.keyMaterial.leanModule !== expectedKeyMaterial.leanModule ||
    envelope.keyMaterial.sourceNodeCount !==
      expectedKeyMaterial.sourceNodeCount ||
    envelope.keyMaterial.ggNodeCount !== expectedKeyMaterial.ggNodeCount ||
    envelope.keyMaterial.ggEdgeCount !== expectedKeyMaterial.ggEdgeCount ||
    envelope.keyMaterial.projectionOperand !==
      expectedKeyMaterial.projectionOperand ||
    envelope.keyMaterial.localOperand !== expectedKeyMaterial.localOperand ||
    envelope.keyMaterial.programChecksum !== expectedKeyMaterial.programChecksum
  ) {
    throw new Error('iota program key material mismatch');
  }
  if (
    protocol69IotaProgramComputeCacheKey(
      envelope.ggSource,
      envelope.session,
      envelope.keyMaterial
    ) !== envelope.computeCacheKey
  ) {
    throw new Error('iota program compute cache key mismatch');
  }
  if (
    envelope.leafCount < 1 ||
    envelope.applicationCount < 0 ||
    envelope.nodeCount !== envelope.leafCount + envelope.applicationCount
  ) {
    throw new Error('iota program structural counts mismatch');
  }
  return envelope;
}

/**
 * Handles the gnosis validate Protocol69 In Room Rf Session workflow.
 */
export function validateProtocol69InRoomRfSession(
  session: Protocol69InRoomRfSession
): Protocol69InRoomRfSession {
  if (session.transport !== 'in-room-rf') {
    throw new Error('protocol69 iota session must be in-room RF');
  }
  if (session.radio !== 'pluto-sdr') {
    throw new Error('protocol69 iota session must target Pluto SDR');
  }
  if (!Number.isInteger(session.antennaCount) || session.antennaCount < 1) {
    throw new Error('protocol69 iota session antennaCount must be positive');
  }
  if (
    session.carrierHz !== undefined &&
    (!Number.isFinite(session.carrierHz) || session.carrierHz <= 0)
  ) {
    throw new Error('protocol69 iota session carrierHz must be positive');
  }
  if (
    session.attenuationDb !== undefined &&
    !Number.isFinite(session.attenuationDb)
  ) {
    throw new Error('protocol69 iota session attenuationDb must be finite');
  }
  return session;
}

/**
 * Handles the gnosis run Protocol69 Iota Program Teleport workflow.
 */
export function runProtocol69IotaProgramTeleport(
  program: IotaTree,
  session: Protocol69InRoomRfSession = protocol69DefaultInRoomRfSession()
): Protocol69IotaProgramCacheReceipt {
  const envelope = validateProtocol69IotaProgramTeleportEnvelope(
    buildProtocol69IotaProgramTeleportEnvelope(program, session)
  );
  const parsed = new BettyCompiler().parse(envelope.ggSource);
  const errors = parsed.diagnostics.filter(
    (diagnostic) => diagnostic.severity === 'error'
  );
  if (errors.length > 0) {
    throw new Error(
      `protocol69 iota GG inflation failed: ${errors
        .map((diagnostic) => diagnostic.message)
        .join('; ')}`
    );
  }
  if (!parsed.ast) {
    throw new Error('protocol69 iota GG inflation produced no topology');
  }

  const recovered = graphAstToIotaTree(parsed.ast);
  const leafCount = iotaTreeLeafCount(recovered);
  const applicationCount = iotaTreeApplicationCount(recovered);
  if (
    leafCount !== envelope.leafCount ||
    applicationCount !== envelope.applicationCount
  ) {
    throw new Error('protocol69 iota inflated topology count drift');
  }
  if (
    parsed.ast.nodes.size !== envelope.keyMaterial.ggNodeCount ||
    parsed.ast.edges.length !== envelope.keyMaterial.ggEdgeCount
  ) {
    throw new Error('protocol69 iota GG key material count drift');
  }

  return {
    schemaVersion: PROTOCOL69_IOTA_PROGRAM_RECEIPT_SCHEMA_VERSION,
    teleportKind: 'iota-program-cache-key',
    session: envelope.session,
    projection: envelope.projection,
    leanModule: envelope.leanModule,
    programChecksum: envelope.programChecksum,
    keyMaterial: envelope.keyMaterial,
    computeCacheKey: envelope.computeCacheKey,
    ggNodeCount: parsed.ast.nodes.size,
    ggEdgeCount: parsed.ast.edges.length,
    leafCount,
    applicationCount,
    nodeCount: leafCount + applicationCount,
  };
}

// ---------------------------------------------------------------------------
// Jet-engine compressor cascade — protocol69 teleport cache-skip stage (OSI L5)
// ---------------------------------------------------------------------------

/**
 * Wire size of a protocol69 cache-key projection envelope, in bytes.
 *
 * SOURCE: the fixed binary layout in `distributed-inference/src/protocol69.rs` /
 * `x-gnosis/gnosis-uring/src/protocol69.rs` — `PROTOCOL69_ENVELOPE_LEN = 32`
 * (4 magic + 4 broadcast + 4 operand + 4 projection + 8 prefix_hash + 4 prefix_len
 *  + 2 layer_lo + 2 layer_hi). This is what crosses the Flow/io_uring wire on a hit.
 */
export const PROTOCOL69_ENVELOPE_BYTES = 32;

/**
 * Bytes of frozen amplituhedron volume a canonical teleport HIT replays LOCALLY
 * (never crosses the wire), in f32 lanes ×4 bytes.
 *
 * SOURCE: `teleport.rs` canonical-hit payload `(tail_residual, kv_slab)` =
 * `(Vec<f32> len 8, Vec<f32> len 64)` as captured in
 * `distributed_inference::teleport::tests::canonical_hit_is_geodesic_zero`
 * and `amplituhedron::AmplituhedronCache::{capture,replay}`:
 *   (8 + 64) f32 × 4 bytes = 288 bytes of compute output served from cache.
 *
 * The teleport stage's COMPRESSION ratio = replayed-output-bytes / wire-key-bytes:
 * the 32-byte key teleports; 288 bytes of frozen compute land for free.
 */
export const PROTOCOL69_TELEPORT_TAIL_RESIDUAL_F32 = 8;
export const PROTOCOL69_TELEPORT_KV_SLAB_F32 = 64;
export const PROTOCOL69_TELEPORT_HIT_OUTPUT_BYTES =
  (PROTOCOL69_TELEPORT_TAIL_RESIDUAL_F32 + PROTOCOL69_TELEPORT_KV_SLAB_F32) * 4; // 288

/**
 * The protocol69 teleport cache-skip ratio (byte source, honest & structural):
 *   replayed frozen-compute output bytes / cache-key wire bytes = 288 / 32 = 9.
 * One 32-byte key admits 288 bytes of frozen amplituhedron compute output —
 * a 9× key-to-payload compression on the L5 session admission boundary.
 */
export const PROTOCOL69_TELEPORT_CACHE_SKIP_RATIO =
  PROTOCOL69_TELEPORT_HIT_OUTPUT_BYTES / PROTOCOL69_ENVELOPE_BYTES; // 9

/**
 * Op-count source for the admission validation stage (honest & structural):
 * `validate_protocol69_projection_envelope` performs exactly ONE XOR plus the
 * fixed equality guards to admit the geodesicLength:0 replay, INSTEAD of the
 * heavy WASM inference kernel that would otherwise run. We count the cheap,
 * provable structural unit — the single XOR projection — as the gate's work,
 * versus the canonical `[2,3,5]` E8-Coxeter compressor that the cached path
 * stands in for downstream. We keep this conservative: the validation gate
 * itself is a clean pass-through (ratio 1, the cascade identity) so the cascade
 * never over-claims; all the compression is the byte cache-skip stage above.
 */
export const PROTOCOL69_ADMISSION_GATE_RATIO = 1;

/**
 * Build the protocol69 teleport cascade as compressor stages.
 *
 * Two honest stages on the L5 (session) teleport-admission boundary:
 *   1. `p69-admission-gate`  — the fail-closed XOR validation (66^7=69). Clean
 *      pass-through (ratio 1 = cascade identity): it gates, it does not compress.
 *   2. `p69-teleport-cache-skip` — the geodesicLength:0 frozen-volume replay.
 *      ratio = HIT output bytes / cache-key wire bytes = 288/32 = 9 (byte source).
 *
 * overall = product = 1 × 9 = 9, per `Gnosis.MathJetEngine.overallRatio`.
 */
export function protocol69CascadeStages(): Stage[] {
  return [
    stage('p69-admission-gate', PROTOCOL69_ADMISSION_GATE_RATIO),
    stage('p69-teleport-cache-skip', PROTOCOL69_TELEPORT_CACHE_SKIP_RATIO),
  ];
}

/**
 * The full protocol69 teleport compressor cascade report. The cache-skip stage is
 * deterministic (288/32 = 9 exactly) so measured end-to-end equals the predicted
 * product — the product law predicts the pipeline exactly (residual 1).
 */
export function protocol69CompressorCascade(): CascadeReport {
  const stages = protocol69CascadeStages();
  const measured = overallRatio(stages); // deterministic byte ratio: measured == predicted
  return cascadeReport('protocol69-teleport[L5/session]', stages, measured);
}
