// @a0n/gnosis/src/protocol69-r1-witness.ts
import {
  buildProtocol69ProjectionEnvelope,
  PROTOCOL69_BROADCAST_SYMBOL,
  PROTOCOL69_EXPECTED_PROJECTION,
  PROTOCOL69_LOCAL_FOIL_OPERAND,
  PROTOCOL69_NAME,
  PROTOCOL69_OPERATOR,
  PROTOCOL69_SCHEMA_VERSION,
  PROTOCOL69_WORD_NAME,
  protocol69Projection,
  type Protocol69ProjectionEnvelope,
} from './protocol69';

export const R1_FOIL_WITNESS_SCHEMA_VERSION =
  'gnosis.protocol69.r1-foil-witness.v1';
export const R1_FOIL_WITNESS_CONTRACT_ID =
  'gnosis.protocol69.r1-foil-witness.btle-channel.v1';
export const R1_FOIL_WITNESS_THEOREM_ID =
  'FoilLoserPolicy.r1_btle_witness_admits_channelized_entropy';

export interface R1FoilWitnessReportInput {
  readonly deviceId: string;
  readonly channelId: number;
  readonly rssiDbm: number;
  readonly packetSeen: boolean;
  readonly cadenceMs?: number;
  readonly timestampMs: number;
  readonly projection?: Protocol69ProjectionEnvelope;
}

export interface R1FoilWitnessReport extends R1FoilWitnessReportInput {
  readonly schemaVersion: typeof R1_FOIL_WITNESS_SCHEMA_VERSION;
  readonly contractId: typeof R1_FOIL_WITNESS_CONTRACT_ID;
  readonly ingress: 'BluetoothChannelWitness';
  readonly theoremId: typeof R1_FOIL_WITNESS_THEOREM_ID;
}

export interface R1FoilChannelWitness {
  readonly ingress: 'BluetoothChannelWitness';
  readonly channelId: number;
  readonly frequencyHz: null;
  readonly rssiDbm: number;
  readonly packetSeen: boolean;
  readonly cadenceMs?: number;
}

function asRecord(value: unknown): Record<string, unknown> | null {
  if (!value || typeof value !== 'object' || Array.isArray(value)) {
    return null;
  }
  return value as Record<string, unknown>;
}

function isFiniteInteger(value: unknown): value is number {
  return (
    typeof value === 'number' &&
    Number.isFinite(value) &&
    Number.isInteger(value)
  );
}

function normalizeDeviceId(value: unknown): string {
  if (typeof value !== 'string') {
    throw new Error('deviceId must be a string');
  }
  const trimmed = value.trim();
  if (!trimmed) {
    throw new Error('deviceId must be nonempty');
  }
  return trimmed;
}

function normalizeChannelId(value: unknown): number {
  if (!isFiniteInteger(value) || value < 0 || value > 0xffff) {
    throw new Error('channelId must be an integer in [0, 65535]');
  }
  return value;
}

function normalizeRssiDbm(value: unknown): number {
  if (!isFiniteInteger(value) || value < -127 || value > 20) {
    throw new Error('rssiDbm must be an integer in [-127, 20]');
  }
  return value;
}

function normalizeTimestampMs(value: unknown): number {
  if (!isFiniteInteger(value) || value < 0) {
    throw new Error('timestampMs must be a nonnegative integer');
  }
  return value;
}

function normalizeCadenceMs(value: unknown): number | undefined {
  if (value === undefined) {
    return undefined;
  }
  if (!isFiniteInteger(value) || value < 0 || value > 0xffff_ffff) {
    throw new Error('cadenceMs must be a nonnegative u32 integer');
  }
  return value;
}

/**
 * Handles the gnosis validate Protocol69 Projection Envelope workflow.
 */
export function validateProtocol69ProjectionEnvelope(
  projection: unknown
): Protocol69ProjectionEnvelope {
  const record = asRecord(projection);
  if (!record) {
    throw new Error('projection must be an object');
  }

  const broadcastSymbol = record['broadcastSymbol'];
  const localFoilOperand = record['localFoilOperand'];
  const expectedProjection = record['expectedProjection'];
  if (
    record['schemaVersion'] !== PROTOCOL69_SCHEMA_VERSION ||
    record['protocol'] !== PROTOCOL69_NAME ||
    record['wordName'] !== PROTOCOL69_WORD_NAME ||
    record['operator'] !== PROTOCOL69_OPERATOR ||
    broadcastSymbol !== PROTOCOL69_BROADCAST_SYMBOL ||
    localFoilOperand !== PROTOCOL69_LOCAL_FOIL_OPERAND ||
    expectedProjection !== PROTOCOL69_EXPECTED_PROJECTION ||
    protocol69Projection(
      broadcastSymbol as number,
      localFoilOperand as number
    ) !== expectedProjection
  ) {
    throw new Error('projection must match canonical protocol69');
  }

  return record as unknown as Protocol69ProjectionEnvelope;
}

/**
 * Builds the R1 Foil Witness Report.
 */
export function buildR1FoilWitnessReport(
  input: R1FoilWitnessReportInput
): R1FoilWitnessReport {
  const cadenceMs = normalizeCadenceMs(input.cadenceMs);
  const projection = input.projection
    ? validateProtocol69ProjectionEnvelope(input.projection)
    : buildProtocol69ProjectionEnvelope();

  return {
    schemaVersion: R1_FOIL_WITNESS_SCHEMA_VERSION,
    contractId: R1_FOIL_WITNESS_CONTRACT_ID,
    ingress: 'BluetoothChannelWitness',
    theoremId: R1_FOIL_WITNESS_THEOREM_ID,
    deviceId: normalizeDeviceId(input.deviceId),
    channelId: normalizeChannelId(input.channelId),
    rssiDbm: normalizeRssiDbm(input.rssiDbm),
    packetSeen: input.packetSeen === true,
    ...(cadenceMs === undefined ? {} : { cadenceMs }),
    timestampMs: normalizeTimestampMs(input.timestampMs),
    projection,
  };
}

/**
 * Handles the gnosis validate R1 Foil Witness Report workflow.
 */
export function validateR1FoilWitnessReport(
  value: unknown
): R1FoilWitnessReport {
  const record = asRecord(value);
  if (!record) {
    throw new Error('R1 FOIL witness report must be an object');
  }
  if (record['schemaVersion'] !== R1_FOIL_WITNESS_SCHEMA_VERSION) {
    throw new Error('schemaVersion mismatch');
  }
  if (record['contractId'] !== R1_FOIL_WITNESS_CONTRACT_ID) {
    throw new Error('contractId mismatch');
  }
  if (record['ingress'] !== 'BluetoothChannelWitness') {
    throw new Error('ingress must be BluetoothChannelWitness');
  }
  if (record['theoremId'] !== R1_FOIL_WITNESS_THEOREM_ID) {
    throw new Error('theoremId mismatch');
  }
  if (typeof record['packetSeen'] !== 'boolean') {
    throw new Error('packetSeen must be boolean');
  }

  const cadenceMs = normalizeCadenceMs(record['cadenceMs']);
  return buildR1FoilWitnessReport({
    deviceId: normalizeDeviceId(record['deviceId']),
    channelId: normalizeChannelId(record['channelId']),
    rssiDbm: normalizeRssiDbm(record['rssiDbm']),
    packetSeen: record['packetSeen'],
    ...(cadenceMs === undefined ? {} : { cadenceMs }),
    timestampMs: normalizeTimestampMs(record['timestampMs']),
    projection: validateProtocol69ProjectionEnvelope(record['projection']),
  });
}

/**
 * Handles the gnosis r1 Foil Report To Channel Witness workflow.
 */
export function r1FoilReportToChannelWitness(
  report: R1FoilWitnessReport
): R1FoilChannelWitness {
  return {
    ingress: 'BluetoothChannelWitness',
    channelId: report.channelId,
    frequencyHz: null,
    rssiDbm: report.rssiDbm,
    packetSeen: report.packetSeen,
    ...(report.cadenceMs === undefined ? {} : { cadenceMs: report.cadenceMs }),
  };
}
