// @a0n/gnosis/src/quantum.ts
/**
 * quantum.ts -- Graph-native quantum values and transitions
 *
 * Qubits, registers, gates, measurement, and collapse as first-class
 * topology primitives. Aligned with OBSERVE, RACE, and FOLD semantics.
 *
 * Roadmap item #5: reuses the value/effect/verification model.
 */

// ============================================================================
// Qubit -- the quantum unit
// ============================================================================

/** Complex number as [real, imaginary] */
export type Complex = [number, number];

/**
 * Handles the gnosis complex workflow.
 */
export function complex(re: number, im: number = 0): Complex {
  return [re, im];
}

/**
 * Handles the gnosis complex Mul workflow.
 */
export function complexMul(a: Complex, b: Complex): Complex {
  return [a[0] * b[0] - a[1] * b[1], a[0] * b[1] + a[1] * b[0]];
}

/**
 * Handles the gnosis complex Add workflow.
 */
export function complexAdd(a: Complex, b: Complex): Complex {
  return [a[0] + b[0], a[1] + b[1]];
}

/**
 * Handles the gnosis complex Norm workflow.
 */
export function complexNorm(a: Complex): number {
  return a[0] * a[0] + a[1] * a[1];
}

/**
 * Handles the gnosis complex Conj workflow.
 */
export function complexConj(a: Complex): Complex {
  return [a[0], -a[1]];
}

function requireComplex(value: Complex | undefined, context: string): Complex {
  if (!value) {
    throw new Error(context);
  }
  return value;
}

/**
 * A qubit: two-dimensional quantum state vector.
 * |ψ⟩ = α|0⟩ + β|1⟩ where |α|² + |β|² = 1
 */
export interface Qubit {
  /** Amplitude for |0⟩ */
  alpha: Complex;
  /** Amplitude for |1⟩ */
  beta: Complex;
}

/**
 * Handles the gnosis qubit workflow.
 */
export function qubit(alpha: Complex, beta: Complex): Qubit {
  return { alpha, beta };
}

/** |0⟩ state */
export function ket0(): Qubit {
  return qubit(complex(1), complex(0));
}

/** |1⟩ state */
export function ket1(): Qubit {
  return qubit(complex(0), complex(1));
}

/** |+⟩ state = (|0⟩ + |1⟩)/√2 */
export function ketPlus(): Qubit {
  const s = 1 / Math.sqrt(2);
  return qubit(complex(s), complex(s));
}

/** |-⟩ state = (|0⟩ - |1⟩)/√2 */
export function ketMinus(): Qubit {
  const s = 1 / Math.sqrt(2);
  return qubit(complex(s), complex(-s));
}

/** Check if qubit state is normalized (|α|² + |β|² ≈ 1) */
export function isNormalized(q: Qubit, tolerance: number = 1e-10): boolean {
  const norm = complexNorm(q.alpha) + complexNorm(q.beta);
  return Math.abs(norm - 1) < tolerance;
}

/** Probability of measuring |0⟩ */
export function prob0(q: Qubit): number {
  return complexNorm(q.alpha);
}

/** Probability of measuring |1⟩ */
export function prob1(q: Qubit): number {
  return complexNorm(q.beta);
}

// ============================================================================
// Quantum Register -- multi-qubit state
// ============================================================================

/**
 * A quantum register: state vector over 2^n basis states.
 * For n qubits, amplitudes has 2^n entries.
 */
export interface QuantumRegister {
  /** Number of qubits */
  numQubits: number;
  /** State vector amplitudes (length 2^numQubits) */
  amplitudes: Complex[];
}

/**
 * Creates the Register.
 */
export function createRegister(numQubits: number): QuantumRegister {
  const size = 1 << numQubits;
  const amplitudes: Complex[] = new Array(size)
    .fill(null)
    .map(() => complex(0));
  amplitudes[0] = complex(1); // Initialize to |00...0⟩
  return { numQubits, amplitudes };
}

/**
 * Handles the gnosis register From Qubit workflow.
 */
export function registerFromQubit(q: Qubit): QuantumRegister {
  return { numQubits: 1, amplitudes: [q.alpha, q.beta] };
}

/**
 * Handles the gnosis register Normalized workflow.
 */
export function registerNormalized(
  reg: QuantumRegister,
  tolerance: number = 1e-10
): boolean {
  const norm = reg.amplitudes.reduce((s, a) => s + complexNorm(a), 0);
  return Math.abs(norm - 1) < tolerance;
}

// ============================================================================
// Gates -- unitary transformations
// ============================================================================

/** 2x2 matrix as [[a,b],[c,d]] of complex numbers */
export type Gate2x2 = [[Complex, Complex], [Complex, Complex]];

function applyGate2x2(q: Qubit, gate: Gate2x2): Qubit {
  const row0 = gate[0];
  const row1 = gate[1];
  if (!row0 || !row1) {
    throw new Error('applyGate2x2: malformed gate');
  }
  const alpha = complexAdd(
    complexMul(requireComplex(row0[0], 'applyGate2x2 gate[0][0]'), q.alpha),
    complexMul(requireComplex(row0[1], 'applyGate2x2 gate[0][1]'), q.beta)
  );
  const beta = complexAdd(
    complexMul(requireComplex(row1[0], 'applyGate2x2 gate[1][0]'), q.alpha),
    complexMul(requireComplex(row1[1], 'applyGate2x2 gate[1][1]'), q.beta)
  );
  return qubit(alpha, beta);
}

/** Hadamard gate: creates superposition */
export const H_GATE: Gate2x2 = (() => {
  const s = 1 / Math.sqrt(2);
  return [
    [complex(s), complex(s)],
    [complex(s), complex(-s)],
  ] as Gate2x2;
})();

/** Pauli-X gate (NOT): |0⟩↔|1⟩ */
export const X_GATE: Gate2x2 = [
  [complex(0), complex(1)],
  [complex(1), complex(0)],
];

/** Pauli-Y gate */
export const Y_GATE: Gate2x2 = [
  [complex(0), [0, -1]],
  [[0, 1], complex(0)],
];

/** Pauli-Z gate: phase flip on |1⟩ */
export const Z_GATE: Gate2x2 = [
  [complex(1), complex(0)],
  [complex(0), complex(-1)],
];

/** S gate (√Z): π/2 phase */
export const S_GATE: Gate2x2 = [
  [complex(1), complex(0)],
  [complex(0), [0, 1]],
];

/** T gate (√S): π/4 phase */
export const T_GATE: Gate2x2 = (() => {
  const s = 1 / Math.sqrt(2);
  return [
    [complex(1), complex(0)],
    [complex(0), complex(s, s)],
  ] as Gate2x2;
})();

/** Apply a single-qubit gate */
export function applyGate(q: Qubit, gate: Gate2x2): Qubit {
  return applyGate2x2(q, gate);
}

/** Apply Hadamard */
export function hadamard(q: Qubit): Qubit {
  return applyGate(q, H_GATE);
}

/** Apply Pauli-X (NOT) */
export function pauliX(q: Qubit): Qubit {
  return applyGate(q, X_GATE);
}

/** Apply Pauli-Z */
export function pauliZ(q: Qubit): Qubit {
  return applyGate(q, Z_GATE);
}

/** Phase rotation gate R(θ) */
export function phaseGate(theta: number): Gate2x2 {
  return [
    [complex(1), complex(0)],
    [complex(0), complex(Math.cos(theta), Math.sin(theta))],
  ];
}

// ============================================================================
// Two-qubit gates on registers
// ============================================================================

/**
 * Apply CNOT gate on a register.
 * control and target are qubit indices (0-based).
 */
export function applyCNOT(
  reg: QuantumRegister,
  control: number,
  target: number
): QuantumRegister {
  const size = 1 << reg.numQubits;
  const newAmps: Complex[] = reg.amplitudes.map(() => complex(0));

  for (let i = 0; i < size; i++) {
    const controlBit = (i >> (reg.numQubits - 1 - control)) & 1;
    if (controlBit === 1) {
      // Flip target bit
      const j = i ^ (1 << (reg.numQubits - 1 - target));
      const existing = newAmps[j];
      const amplitude = reg.amplitudes[i];
      if (!existing || !amplitude) {
        throw new Error('applyCNOT: missing amplitude');
      }
      newAmps[j] = complexAdd(existing, amplitude);
    } else {
      const existing = newAmps[i];
      const amplitude = reg.amplitudes[i];
      if (!existing || !amplitude) {
        throw new Error('applyCNOT: missing amplitude');
      }
      newAmps[i] = complexAdd(existing, amplitude);
    }
  }

  return { numQubits: reg.numQubits, amplitudes: newAmps };
}

/**
 * Apply CZ (controlled-Z) gate on a register.
 */
export function applyCZ(
  reg: QuantumRegister,
  control: number,
  target: number
): QuantumRegister {
  const size = 1 << reg.numQubits;
  const newAmps: Complex[] = [...reg.amplitudes.map((a) => [...a] as Complex)];

  for (let i = 0; i < size; i++) {
    const controlBit = (i >> (reg.numQubits - 1 - control)) & 1;
    const targetBit = (i >> (reg.numQubits - 1 - target)) & 1;
    if (controlBit === 1 && targetBit === 1) {
      const amplitude = newAmps[i];
      if (!amplitude) {
        throw new Error('applyCZ: missing amplitude');
      }
      newAmps[i] = complexMul(amplitude, complex(-1));
    }
  }

  return { numQubits: reg.numQubits, amplitudes: newAmps };
}

/**
 * Apply a single-qubit gate to a specific qubit in a register.
 */
export function applyGateToRegister(
  reg: QuantumRegister,
  qubitIdx: number,
  gate: Gate2x2
): QuantumRegister {
  const size = 1 << reg.numQubits;
  const newAmps: Complex[] = new Array(size).fill(null).map(() => complex(0));

  for (let i = 0; i < size; i++) {
    const bit = (i >> (reg.numQubits - 1 - qubitIdx)) & 1;
    const partner = i ^ (1 << (reg.numQubits - 1 - qubitIdx));
    const inputAmplitude = reg.amplitudes[i];
    const partnerAmplitude = reg.amplitudes[partner];
    const resultAmplitude = newAmps[i];

    if (!inputAmplitude || !partnerAmplitude || !resultAmplitude) {
      throw new Error('applyGateToRegister: missing amplitude');
    }

    if (bit === 0) {
      // |0⟩ component
      let outputAmplitude = resultAmplitude;
      outputAmplitude = complexAdd(
        outputAmplitude,
        complexMul(requireComplex(gate[0][0], 'applyGateToRegister gate[0][0]'), inputAmplitude)
      );
      outputAmplitude = complexAdd(
        outputAmplitude,
        complexMul(requireComplex(gate[0][1], 'applyGateToRegister gate[0][1]'), partnerAmplitude)
      );
      newAmps[i] = outputAmplitude;
    } else {
      // |1⟩ component
      let outputAmplitude = resultAmplitude;
      outputAmplitude = complexAdd(
        outputAmplitude,
        complexMul(requireComplex(gate[1][0], 'applyGateToRegister gate[1][0]'), partnerAmplitude)
      );
      outputAmplitude = complexAdd(
        outputAmplitude,
        complexMul(requireComplex(gate[1][1], 'applyGateToRegister gate[1][1]'), inputAmplitude)
      );
      newAmps[i] = outputAmplitude;
    }
  }

  return { numQubits: reg.numQubits, amplitudes: newAmps };
}

// ============================================================================
// Measurement -- collapse aligned with OBSERVE semantics
// ============================================================================

export interface MeasurementOutcome {
  /** Classical bit result (0 or 1) */
  result: 0 | 1;
  /** Probability of this outcome */
  probability: number;
  /** Post-measurement qubit state */
  postState: Qubit;
}

/**
 * Measure a qubit, collapsing its state.
 * Returns the outcome and post-measurement state.
 * rng provides randomness (like void walker's rng parameter).
 */
export function measure(q: Qubit, rng: () => number): MeasurementOutcome {
  const p0 = prob0(q);
  const r = rng();

  if (r < p0) {
    return {
      result: 0,
      probability: p0,
      postState: ket0(),
    };
  } else {
    return {
      result: 1,
      probability: 1 - p0,
      postState: ket1(),
    };
  }
}

/**
 * Measure a specific qubit in a register.
 * Returns the classical result and the collapsed register.
 */
export function measureRegister(
  reg: QuantumRegister,
  qubitIdx: number,
  rng: () => number
): { result: 0 | 1; register: QuantumRegister } {
  const size = 1 << reg.numQubits;

  // Calculate probability of measuring 0
  let p0 = 0;
  for (let i = 0; i < size; i++) {
    const bit = (i >> (reg.numQubits - 1 - qubitIdx)) & 1;
    if (bit === 0) {
      const amplitude = reg.amplitudes[i];
      if (!amplitude) {
        throw new Error('measureRegister: missing amplitude');
      }
      p0 += complexNorm(amplitude);
    }
  }

  const result: 0 | 1 = rng() < p0 ? 0 : 1;
  const norm = result === 0 ? Math.sqrt(p0) : Math.sqrt(1 - p0);

  // Collapse: zero out amplitudes inconsistent with result, renormalize
  const newAmps: Complex[] = new Array(size).fill(null).map(() => complex(0));
  for (let i = 0; i < size; i++) {
    const bit = (i >> (reg.numQubits - 1 - qubitIdx)) & 1;
    if (bit === result) {
      const amplitude = reg.amplitudes[i];
      if (!amplitude) {
        throw new Error('measureRegister: missing amplitude');
      }
      newAmps[i] =
        norm > 0
          ? [amplitude[0] / norm, amplitude[1] / norm]
          : complex(0);
    }
  }

  return {
    result,
    register: { numQubits: reg.numQubits, amplitudes: newAmps },
  };
}

// ============================================================================
// Entanglement -- Bell states and detection
// ============================================================================

/** Create a Bell state |Φ+⟩ = (|00⟩ + |11⟩)/√2 */
export function bellState(): QuantumRegister {
  const s = 1 / Math.sqrt(2);
  return {
    numQubits: 2,
    amplitudes: [complex(s), complex(0), complex(0), complex(s)],
  };
}

/**
 * Check if a two-qubit register is entangled (non-separable).
 * Uses the concurrence measure.
 */
export function isEntangled(
  reg: QuantumRegister,
  tolerance: number = 1e-10
): boolean {
  if (reg.numQubits !== 2) return false;
  const a = reg.amplitudes[0];
  const b = reg.amplitudes[1];
  const c = reg.amplitudes[2];
  const d = reg.amplitudes[3];
  if (!a || !b || !c || !d) {
    return false;
  }
  // Concurrence = 2|ad - bc|
  const adRe = a[0] * d[0] - a[1] * d[1];
  const adIm = a[0] * d[1] + a[1] * d[0];
  const bcRe = b[0] * c[0] - b[1] * c[1];
  const bcIm = b[0] * c[1] + b[1] * c[0];
  const diffRe = adRe - bcRe;
  const diffIm = adIm - bcIm;
  const concurrence = 2 * Math.sqrt(diffRe * diffRe + diffIm * diffIm);
  return concurrence > tolerance;
}

// ============================================================================
// Quantum circuit -- sequence of operations for topology execution
// ============================================================================

export type CircuitOp =
  | { type: 'gate'; qubit: number; gate: Gate2x2; name: string }
  | { type: 'cnot'; control: number; target: number }
  | { type: 'cz'; control: number; target: number }
  | { type: 'measure'; qubit: number }
  | { type: 'barrier' };

export interface QuantumCircuit {
  numQubits: number;
  ops: CircuitOp[];
}

/**
 * Creates the Circuit.
 */
export function createCircuit(numQubits: number): QuantumCircuit {
  return { numQubits, ops: [] };
}

/**
 * Handles the gnosis add Gate workflow.
 */
export function addGate(
  circuit: QuantumCircuit,
  qubit: number,
  gate: Gate2x2,
  name: string
): void {
  circuit.ops.push({ type: 'gate', qubit, gate, name });
}

/**
 * Handles the gnosis add CNOT workflow.
 */
export function addCNOT(
  circuit: QuantumCircuit,
  control: number,
  target: number
): void {
  circuit.ops.push({ type: 'cnot', control, target });
}

/**
 * Handles the gnosis add CZ workflow.
 */
export function addCZ(
  circuit: QuantumCircuit,
  control: number,
  target: number
): void {
  circuit.ops.push({ type: 'cz', control, target });
}

/**
 * Handles the gnosis add Measure workflow.
 */
export function addMeasure(circuit: QuantumCircuit, qubit: number): void {
  circuit.ops.push({ type: 'measure', qubit });
}

/**
 * Handles the gnosis add Barrier workflow.
 */
export function addBarrier(circuit: QuantumCircuit): void {
  circuit.ops.push({ type: 'barrier' });
}

/**
 * Execute a quantum circuit and return measurement results.
 */
export function executeCircuit(
  circuit: QuantumCircuit,
  rng: () => number
): { register: QuantumRegister; measurements: Map<number, 0 | 1> } {
  let reg = createRegister(circuit.numQubits);
  const measurements = new Map<number, 0 | 1>();

  for (const op of circuit.ops) {
    switch (op.type) {
      case 'gate':
        reg = applyGateToRegister(reg, op.qubit, op.gate);
        break;
      case 'cnot':
        reg = applyCNOT(reg, op.control, op.target);
        break;
      case 'cz':
        reg = applyCZ(reg, op.control, op.target);
        break;
      case 'measure': {
        const outcome = measureRegister(reg, op.qubit, rng);
        reg = outcome.register;
        measurements.set(op.qubit, outcome.result);
        break;
      }
      case 'barrier':
        break; // Synchronization point, no-op in simulation
    }
  }

  return { register: reg, measurements };
}
