// @a0n/gnosis/src/randomness.ts
/**
 * Gnosis **randomness core** — ties the TypeScript simulation PRNG to the Lean
 * formal surface (negligible advantage, `CsPrngSecurity`, games, clinamen
 * anti-bridge). This module does not implement a CSPRNG; it documents contract
 * and re-exports the nested Lacey PRNG used for procedural / scheduling lanes.
 *
 * Human map: `docs/RANDOMNESS.md`. Package import: `@a0n/gnosis/randomness`
 * or root `@a0n/gnosis` (re-exported from `lib.ts`).
 *
 * @module @a0n/gnosis/randomness
 */

/** Stable Lean library paths for tooling, logs, and cross-repo trace anchors. */
export const GNOSIS_RANDOMNESS_LEAN_MODULES = [
  'ForkRaceFoldTheorems.CryptographicPseudorandom',
  'ForkRaceFoldTheorems.CryptographicExternalAxioms',
  'ForkRaceFoldTheorems.FiniteIndPrgGames',
  'ForkRaceFoldTheorems.FiniteIndPrfGames',
  'ForkRaceFoldTheorems.ClinamenSeedIndistinguishability',
  'ForkRaceFoldTheorems.ClinamenCryptoAntiBridge',
  'ForkRaceFoldTheorems.EpistemicRandomness',
  'ForkRaceFoldTheorems.LandauerInformationPhysics',
  'ForkRaceFoldTheorems.BlackHoleInformationSketch',
  'ForkRaceFoldTheorems.AperiodicPseudorandom',
  'ForkRaceFoldTheorems.LaceyNestedRotation',
] as const;

export type GnosisRandomnessLeanModule =
  (typeof GNOSIS_RANDOMNESS_LEAN_MODULES)[number];

/**
 * What this **runtime** layer is allowed to claim. Cryptographic indistinguishability
 * is defined and (when assumed) axiomatized in Lean — not reproduced here.
 */
export type GnosisRandomnessRuntimeStance =
  | 'simulation_prng'
  | 'formal_csprng_claim_requires_lean_extrinsic_axioms';

export const GNOSIS_RANDOMNESS_RUNTIME_STANCE: GnosisRandomnessRuntimeStance =
  'simulation_prng';

/**
 * Short string for logs, manifests, or UI when a consumer must not confuse
 * procedural randomness with cryptographic security.
 */
export const GNOSIS_SIMULATION_PRNG_DISCLAIMER =
  'Gnosis nested Lacey PRNG is simulation-grade only; CsPrngSecurity and negligible advantage are formalized in Lean (ForkRaceFoldTheorems.CryptographicPseudorandom), not asserted by this TypeScript runtime.';

export {
  createLaceyNestedRng,
  HELIX_TURN_SAMPLES,
  isAlphaHalf,
  LaceyNestedRng,
  rotationAt,
} from './lacey-nested-prng';
