// @a0n/gnosis/src/reduction-projection-scan.ts
/**
 * Cross-reduction projection scanner.
 *
 * Adapts multiple reduction surfaces into the shared projection-paradox
 * shape/unit contract from `spiderweb-topology.ts`, so we can compare
 * predicted vs observed paradox units across domains.
 */

import {
  assessVoidDimension,
  type GnosisVoidDimensionSignal,
} from './void-dimensions';
import {
  recommendUntanglingStrategy,
  type GnosisUntanglingSignal,
} from './untangling-strategy';
import {
  calculateContagionRadius,
  type ContagionVector,
} from './topological-contagion';
import { type FlowObstructionReport } from './flow-obstruction';
import {
  buildTriangulationCoercionMatrix,
  coercionEdgeToReductionInput,
  type TriMesh,
} from './triangulation-dimensional-coercion';
import {
  type MeshConfig,
  type ReductionProjectionInput,
  type ReductionProjectionScanReport,
  scanReductionProjections,
} from './spiderweb-topology';

export interface VoidReductionProjectionSignal
  extends GnosisVoidDimensionSignal {
  readonly id?: string;
  readonly observedDim?: number;
}

export interface UntanglingReductionProjectionSignal
  extends GnosisUntanglingSignal {
  readonly id?: string;
  readonly observedDim?: number;
}

export interface ContagionReductionProjectionSignal {
  readonly id?: string;
  readonly mesh: MeshConfig;
  readonly vector: ContagionVector;
  readonly observedDim?: number;
}

export interface ObstructionReductionProjectionSignal {
  readonly id?: string;
  readonly report: FlowObstructionReport;
  readonly observedDimFloor?: number;
}

export interface TriangulationReductionProjectionSignal {
  readonly idPrefix?: string;
  readonly mesh: TriMesh;
  readonly dimensions: readonly number[];
}

export interface CoreReductionScanRequest {
  readonly voidSignals?: readonly VoidReductionProjectionSignal[];
  readonly untanglingSignals?: readonly UntanglingReductionProjectionSignal[];
  readonly contagionSignals?: readonly ContagionReductionProjectionSignal[];
  readonly obstructionSignals?: readonly ObstructionReductionProjectionSignal[];
  readonly triangulationSignals?: readonly TriangulationReductionProjectionSignal[];
}

export interface CoreReductionScanResult {
  readonly reductions: readonly ReductionProjectionInput[];
  readonly scans: readonly ReductionProjectionScanReport[];
}

function toNat(value: number | undefined | null, fallback: number): number {
  if (!Number.isFinite(value ?? Number.NaN)) return fallback;
  return Math.max(0, Math.trunc(value ?? fallback));
}

/**
 * Builds the Void Reduction Projection Input.
 */
export function buildVoidReductionProjectionInput(
  signal: VoidReductionProjectionSignal
): ReductionProjectionInput {
  const assessment = assessVoidDimension(signal);
  const observedDim = Math.max(1, toNat(signal.observedDim, 2));
  const latentDim = observedDim + Math.max(1, assessment.invisibleCrossings);
  return {
    id: signal.id ?? `void:${signal.dimension}`,
    frame: { observedDim, latentDim },
    witness: {
      observedHasSolution:
        !assessment.gapIsLowerBound && assessment.visibleCrossings > 0,
      latentHasSolution: assessment.invisibleCrossings > 0,
      hasCollapsedSplit:
        assessment.invisibleCrossings > 0 && assessment.gapIsLowerBound,
    },
  };
}

/**
 * Builds the Untangling Reduction Projection Input.
 */
export function buildUntanglingReductionProjectionInput(
  signal: UntanglingReductionProjectionSignal
): ReductionProjectionInput {
  const recommendation = recommendUntanglingStrategy(signal);
  const needsExpansion =
    recommendation.action === 'expand-first' ||
    recommendation.action === 'staged-expand-first';
  const observedDim = Math.max(1, toNat(signal.observedDim, 2));
  const latentDim =
    observedDim +
    (needsExpansion ? Math.max(1, recommendation.unknottingBudget) : 0);
  return {
    id: signal.id ?? `untangling:${recommendation.action}`,
    frame: { observedDim, latentDim },
    witness: {
      observedHasSolution: !needsExpansion,
      latentHasSolution:
        recommendation.unknottingBudget > 0 || recommendation.action === 'hold',
      hasCollapsedSplit:
        needsExpansion && recommendation.unknottingBudget > 0,
    },
  };
}

/**
 * Builds the Contagion Reduction Projection Input.
 */
export function buildContagionReductionProjectionInput(
  signal: ContagionReductionProjectionSignal
): ReductionProjectionInput {
  const observedRadius = calculateContagionRadius(
    signal.mesh,
    signal.vector,
    false
  );
  const latentRadius = calculateContagionRadius(signal.mesh, signal.vector, true);
  const observedDim = Math.max(1, toNat(signal.observedDim, 2));
  const latentDim = observedDim + (latentRadius > observedRadius ? 1 : 0);
  return {
    id: signal.id ?? `contagion:${signal.vector.payloadId}`,
    frame: { observedDim, latentDim },
    witness: {
      observedHasSolution: observedRadius >= latentRadius,
      latentHasSolution: latentRadius >= observedRadius,
      hasCollapsedSplit: latentRadius > observedRadius,
    },
  };
}

/**
 * Builds the Obstruction Reduction Projection Input.
 */
export function buildObstructionReductionProjectionInput(
  signal: ObstructionReductionProjectionSignal
): ReductionProjectionInput {
  const observedDim = Math.max(
    1,
    toNat(signal.observedDimFloor, signal.report.effectiveParallelism)
  );
  const latentDim = observedDim + Math.max(0, toNat(signal.report.totalCrossings, 0));
  const hasUntieMove = signal.report.obstructions.some(
    (obstruction) => obstruction.unknottingNumber > 0
  );
  return {
    id: signal.id ?? `obstruction:${signal.report.functionName}`,
    frame: { observedDim, latentDim },
    witness: {
      observedHasSolution: signal.report.totalCrossings === 0,
      latentHasSolution:
        signal.report.totalCrossings === 0 ? true : hasUntieMove,
      hasCollapsedSplit:
        signal.report.totalCrossings > 0 && hasUntieMove,
    },
  };
}

/**
 * Builds the Triangulation Reduction Projection Inputs.
 */
export function buildTriangulationReductionProjectionInputs(
  signal: TriangulationReductionProjectionSignal
): ReductionProjectionInput[] {
  const matrix = buildTriangulationCoercionMatrix(
    signal.mesh,
    signal.dimensions,
    signal.idPrefix ?? 'triangulation'
  );
  return matrix.edges.map((edge) => coercionEdgeToReductionInput(edge));
}

/**
 * Handles the gnosis collect Core Reduction Projection Inputs workflow.
 */
export function collectCoreReductionProjectionInputs(
  request: CoreReductionScanRequest
): ReductionProjectionInput[] {
  const reductions: ReductionProjectionInput[] = [];
  for (const signal of request.voidSignals ?? []) {
    reductions.push(buildVoidReductionProjectionInput(signal));
  }
  for (const signal of request.untanglingSignals ?? []) {
    reductions.push(buildUntanglingReductionProjectionInput(signal));
  }
  for (const signal of request.contagionSignals ?? []) {
    reductions.push(buildContagionReductionProjectionInput(signal));
  }
  for (const signal of request.obstructionSignals ?? []) {
    reductions.push(buildObstructionReductionProjectionInput(signal));
  }
  for (const signal of request.triangulationSignals ?? []) {
    reductions.push(...buildTriangulationReductionProjectionInputs(signal));
  }
  return reductions;
}

/**
 * Handles the gnosis scan Core Reduction Families workflow.
 */
export function scanCoreReductionFamilies(
  request: CoreReductionScanRequest
): CoreReductionScanResult {
  const reductions = collectCoreReductionProjectionInputs(request);
  return {
    reductions,
    scans: scanReductionProjections(reductions),
  };
}
