// @a0n/gnosis/src/reduction-projection-formalization.ts
import {
  scanCoreReductionFamilies,
  type CoreReductionScanRequest,
  type CoreReductionScanResult,
} from './reduction-projection-scan';

export const REDUCTION_PROJECTION_THEOREM_IDS = [
  'THM-PROJECTION-PARADOX-UNIT-SANDWICH',
  'THM-PROJECTION-PARADOX-PREDICTIVE-EXACT',
  'THM-CROSS-REDUCTION-PROJECTION-FOLD',
  'THM-TRIANGULATION-COERCION-MATRIX-2-3-6-54',
  'THM-SPHERE-TRIANGULATION-BIDIRECTIONAL-ROUNDTRIP',
  'THM-TRIANGULATION-STRICT-UNKNOT-REDUCTION',
] as const;

export type ReductionProjectionTheoremId =
  (typeof REDUCTION_PROJECTION_THEOREM_IDS)[number];

export interface ReductionProjectionFormalizationInput {
  readonly request: CoreReductionScanRequest;
  readonly generatedAt?: string;
  readonly source?: string;
}

export interface ReductionProjectionFormalizationEntry {
  readonly id: string;
  readonly paradox: boolean;
  readonly predictedLowerBound: number;
  readonly observedUnits: number;
  readonly residual: number;
  readonly withinSandwichBounds: boolean;
}

export interface ReductionProjectionFormalizationAggregate {
  readonly reductions: number;
  readonly paradoxCount: number;
  readonly sandwichSatisfiedCount: number;
  readonly predictiveConsistentCount: number;
}

export interface ReductionProjectionFormalizationBundle {
  readonly generatedAt: string;
  readonly source: string;
  readonly theoremIds: readonly ReductionProjectionTheoremId[];
  readonly entries: readonly ReductionProjectionFormalizationEntry[];
  readonly aggregate: ReductionProjectionFormalizationAggregate;
  readonly scan: CoreReductionScanResult;
}

function nowIso(): string {
  return new Date().toISOString();
}

/**
 * Handles the gnosis derive Reduction Projection Formalization workflow.
 */
export function deriveReductionProjectionFormalization(
  input: ReductionProjectionFormalizationInput
): ReductionProjectionFormalizationBundle {
  const scan = scanCoreReductionFamilies(input.request);
  const entries = scan.scans.map<ReductionProjectionFormalizationEntry>((entry) => ({
    id: entry.id,
    paradox: entry.shapeUnit.isProjectionParadox,
    predictedLowerBound: entry.predictive.predictedLowerBound,
    observedUnits: entry.predictive.observedUnits,
    residual: entry.predictive.residual,
    withinSandwichBounds: entry.sandwich.withinBounds,
  }));

  const aggregate: ReductionProjectionFormalizationAggregate = {
    reductions: entries.length,
    paradoxCount: entries.filter((entry) => entry.paradox).length,
    sandwichSatisfiedCount: entries.filter((entry) => entry.withinSandwichBounds)
      .length,
    predictiveConsistentCount: entries.filter((entry) => entry.residual === 0)
      .length,
  };

  return {
    generatedAt: input.generatedAt ?? nowIso(),
    source: input.source ?? 'reductionProjectionFormalization/v1',
    theoremIds: REDUCTION_PROJECTION_THEOREM_IDS,
    entries,
    aggregate,
    scan,
  };
}
