// @a0n/gnosis/src/repl.tsx
import React, { useEffect, useMemo, useRef, useState } from '@a0n/raect';
import { render, Text, Box, useApp, TextInput } from '@a0n/wretch';
import type { WretchColor } from '@a0n/wretch';
import {
  BettyCompiler,
  type ASTEdge,
  type Diagnostic,
  type GraphAST,
} from './betty/compiler.js';
import { GnosisEngine } from './runtime/engine.js';
import { GnosisRegistry } from './runtime/registry.js';
import {
  buildVisualizationFromExecution,
  createInitialVisualization,
  type ReplVisualizationState,
  type VisualNodeStatus,
} from './repl-visualization.js';
import {
  analyzeGnosisSource,
  type GnosisComplexityReport,
  DEFAULT_GNOSIS_STEERING_MODE,
  finishSteeringTelemetry,
  startSteeringTelemetry,
  surfaceSteeringRecommendations,
  surfaceSteeringMetrics,
  withSteeringTelemetry,
  type GnosisSteeringMode,
  type GnosisSteeringMetrics,
} from './analysis.js';
import {
  authorizeBoundaryWalkRun,
  createSteeringTraceRecord,
  DEFAULT_GNOSIS_STEERING_TRACE_WINDOW,
  defaultSteeringTraceRoomName,
  fingerprintSteeringTopology,
  formatSteeringTraceAlert,
  formatSteeringTraceStoreStatus,
  formatSteeringTraceSummary,
  GnosisSteeringTraceStore,
} from './steering-trace.js';
import type { QDocRelayConfig } from './crdt/index.js';
import { authorizeSteeringApply } from './auth/index.js';
import { bootstrapExecutionAuthFromTopology } from './auth/index.js';
import type { GnosisExecutionAuthContext } from './auth/index.js';

type HistoryEntry = {
  type: 'input' | 'output' | 'error';
  text: string;
};

type ReplOptions = {
  verbose?: boolean;
  steeringMode?: GnosisSteeringMode;
  steeringTraceWindow?: number;
  steeringRelayConfig?: QDocRelayConfig | null;
  executionAuth?: GnosisExecutionAuthContext | null;
};

const MAX_HISTORY_ENTRIES = 80;
const MAX_DISPLAYED_HISTORY = 12;
const MAX_SOURCE_LINES = 8;
const BREATH_INTERVAL_MS = 420;
const ERROR_MOTTO = "Baby's Got Stack";
const VERBOSE_MOTTO =
  "Look at the size of that topology... let's see what you've done here.";
const STATUS_STYLE: Record<
  VisualNodeStatus,
  { glyph: string; color: WretchColor }
> = {
  pending: { glyph: '[ ]', color: 'gray' },
  active: { glyph: '[*]', color: 'cyan' },
  completed: { glyph: '[x]', color: 'green' },
  cycled: { glyph: '[!]', color: 'red' },
};
const COLOR_SPECTRUM = [
  'red',
  'yellow',
  'green',
  'cyan',
  'blue',
  'magenta',
  'redBright',
  'yellowBright',
  'greenBright',
  'cyanBright',
  'blueBright',
  'magentaBright',
] as const;
const EDGE_TYPE_COLOR_INDEX: Record<string, number> = {
  PROCESS: 2,
  FORK: 5,
  RACE: 0,
  FOLD: 3,
  COLLAPSE: 3,
  SUPERPOSE: 7,
  EVOLVE: 1,
  ENTANGLE: 10,
  OBSERVE: 4,
  MEASURE: 9,
  HALT: 8,
  VENT: 6,
  TUNNEL: 11,
};

type SpectrumColor = (typeof COLOR_SPECTRUM)[number];

function formatError(error: unknown): string {
  if (error instanceof Error) {
    return error.message;
  }
  return String(error);
}

function formatMottoError(error: unknown): string {
  return `${ERROR_MOTTO}: ${formatError(error)}`;
}

function assertBoundaryWalkAuthorized(
  source: string,
  steeringMode: GnosisSteeringMode,
  executionAuth: GnosisExecutionAuthContext | null
): void {
  if (!executionAuth || executionAuth.enforce !== true) {
    return;
  }

  const runAuthorization = authorizeBoundaryWalkRun({
    source,
    executionAuth,
  });
  if (!runAuthorization.allowed) {
    throw new Error(
      `Boundary walk denied: ${
        runAuthorization.reason ?? 'missing gnosis/steering.run capability'
      }`
    );
  }

  if (steeringMode !== 'apply') {
    return;
  }

  const applyAuthorization = authorizeSteeringApply({
    topologyFingerprint: fingerprintSteeringTopology(source),
    auth: executionAuth,
  });
  if (!applyAuthorization.allowed) {
    throw new Error(
      `Boundary walk apply denied: ${
        applyAuthorization.reason ?? 'missing gnosis/steering.apply capability'
      }`
    );
  }
}

async function resolveBoundaryWalkExecutionAuth(
  ast: GraphAST,
  explicitExecutionAuth: GnosisExecutionAuthContext | null
): Promise<GnosisExecutionAuthContext | null> {
  if (explicitExecutionAuth) {
    return explicitExecutionAuth;
  }

  const bootstrap = await bootstrapExecutionAuthFromTopology(ast, {
    initialPayload: 'REPL_INIT',
  });
  return bootstrap.executionAuth;
}

function formatDiagnostics(diagnostics: Diagnostic[]): string {
  const noteworthy = diagnostics.filter(
    (diagnostic) =>
      diagnostic.severity === 'error' || diagnostic.severity === 'warning'
  );
  if (noteworthy.length === 0) {
    return '';
  }

  const rendered = noteworthy
    .slice(0, 5)
    .map(
      (diagnostic) =>
        `[${diagnostic.severity.toUpperCase()} L${diagnostic.line}:C${
          diagnostic.column
        }] ${diagnostic.message}`
    );
  if (noteworthy.length > rendered.length) {
    rendered.push(
      `... ${noteworthy.length - rendered.length} more diagnostics`
    );
  }

  return rendered.join('\n');
}

function formatCompilationMessage(
  output: string,
  diagnostics: Diagnostic[]
): string {
  const diagnosticSummary = formatDiagnostics(diagnostics);
  if (diagnosticSummary.length === 0) {
    return output;
  }
  return `${output}\n${diagnosticSummary}`;
}

function spectrumColor(index: number, phase: number): SpectrumColor {
  const cursor = (index + Math.floor(phase / 2)) % COLOR_SPECTRUM.length;
  const wrappedCursor = cursor < 0 ? cursor + COLOR_SPECTRUM.length : cursor;
  return COLOR_SPECTRUM[wrappedCursor];
}

function rainbowColor(index: number): SpectrumColor {
  const cursor = index % COLOR_SPECTRUM.length;
  const wrappedCursor = cursor < 0 ? cursor + COLOR_SPECTRUM.length : cursor;
  return COLOR_SPECTRUM[wrappedCursor];
}

function breatheIntensity(phase: number): number {
  return (Math.sin(phase / 8) + 1) / 2;
}

function isDimBreath(phase: number): boolean {
  return breatheIntensity(phase) < 0.45;
}

function edgeColor(edge: ASTEdge): SpectrumColor {
  const offset = EDGE_TYPE_COLOR_INDEX[edge.type] ?? 0;
  return rainbowColor(offset);
}

function edgeTouchesWave(edge: ASTEdge, activeWave: Set<string>): boolean {
  return (
    edge.sourceIds.some((nodeId) => activeWave.has(nodeId.trim())) ||
    edge.targetIds.some((nodeId) => activeWave.has(nodeId.trim()))
  );
}

const SpectrumHeading = React.memo(function SpectrumHeading({
  phase,
  verbose,
}: {
  phase: number;
  verbose: boolean;
}) {
  const title = 'Gnosis Visual Graph REPL v1.3.0';
  const subtitle = verbose
    ? VERBOSE_MOTTO
    : "Baby's Got Stack. Draw topology, execute flow, collapse with style.";

  return (
    <Box flexDirection="column" marginBottom={1}>
      <Text bold={!isDimBreath(phase)} dimColor={isDimBreath(phase)}>
        {title.split('').map((character, index) => (
          <Text key={`title-${index}`} color={spectrumColor(index, phase)}>
            {character}
          </Text>
        ))}
      </Text>
      <Text dimColor={isDimBreath(phase)}>
        {subtitle.split('').map((character, index) => (
          <Text
            key={`subtitle-${index}`}
            color={spectrumColor(index + 2, phase)}
          >
            {character}
          </Text>
        ))}
      </Text>
    </Box>
  );
});

const SourcePanel = React.memo(function SourcePanel({
  sourceLines,
}: {
  sourceLines: string[];
}) {
  if (sourceLines.length === 0) {
    return (
      <Box
        borderStyle="single"
        borderColor={rainbowColor(2)}
        paddingX={1}
        marginY={1}
      >
        <Text color={rainbowColor(3)}>
          Editor empty. Type GGL lines to build the topology.
        </Text>
      </Box>
    );
  }

  const startLine = Math.max(0, sourceLines.length - MAX_SOURCE_LINES);
  const visibleLines = sourceLines.slice(startLine);

  return (
    <Box
      flexDirection="column"
      borderStyle="single"
      borderColor={rainbowColor(1)}
      paddingX={1}
      marginY={1}
    >
      <Text bold color={rainbowColor(0)}>
        TOPOLOGY SOURCE
      </Text>
      {visibleLines.map((line: string, index: number) => {
        const lineNumber = startLine + index + 1;
        return (
          <Box key={`source-${lineNumber}`} flexDirection="row">
            <Text color={rainbowColor(index + 8)}>
              {String(lineNumber).padStart(2, '0')} |{' '}
            </Text>
            <Text color={rainbowColor(index + 4)}>{line}</Text>
          </Box>
        );
      })}
    </Box>
  );
});

const GraphCanvas = React.memo(function GraphCanvas({
  ast,
  visualization,
}: {
  ast: GraphAST | null;
  visualization: ReplVisualizationState;
}) {
  if (!ast || ast.edges.length === 0) {
    return (
      <Box
        borderStyle="round"
        paddingX={1}
        marginY={1}
        borderColor={rainbowColor(6)}
      >
        <Text color={rainbowColor(7)}>
          No topology defined. Start drawing: (a)-[:FORK]-{'>'}(b|c)
        </Text>
      </Box>
    );
  }

  const nodeIds = Array.from(ast.nodes.keys());
  const activeWave = new Set(visualization.activeWaveFunction);

  return (
    <Box
      flexDirection="column"
      borderStyle="double"
      borderColor={rainbowColor(5)}
      paddingX={1}
      marginY={1}
    >
      <Text bold color={rainbowColor(4)}>
        GNOSIS VISUAL GRAPH EDITOR
      </Text>

      <Box flexDirection="column" marginTop={1}>
        <Text bold color={rainbowColor(9)}>
          Nodes
        </Text>
        {nodeIds.map((nodeId: string, index: number) => {
          const style =
            STATUS_STYLE[visualization.nodeStates[nodeId] ?? 'pending'];
          const label = ast.nodes.get(nodeId)?.labels[0];
          return (
            <Box key={nodeId} flexDirection="row">
              <Text color={style.color}>{style.glyph} </Text>
              <Text color={rainbowColor(index)}>{nodeId}</Text>
              {label ? (
                <Text color={rainbowColor(index + 2)}>:{label}</Text>
              ) : null}
            </Box>
          );
        })}
      </Box>

      <Box flexDirection="column" marginTop={1}>
        <Text bold color={rainbowColor(10)}>
          Edges
        </Text>
        {ast.edges.map((edge, index) => {
          const lineColor = edgeColor(edge);
          const isWaveEdge = edgeTouchesWave(edge, activeWave);
          return (
            <Text key={`edge-${index}`} color={lineColor} bold={isWaveEdge}>
              ({edge.sourceIds.join('|')}) -[{edge.type}]-{'>'} (
              {edge.targetIds.join('|')})
            </Text>
          );
        })}
      </Box>

      <Box flexDirection="column" marginTop={1}>
        <Text color={rainbowColor(11)}>
          psi(t):{' '}
          {visualization.activeWaveFunction.length > 0
            ? `[${visualization.activeWaveFunction.join(', ')}]`
            : '[collapsed]'}
        </Text>
        {visualization.quantumEvents.length === 0 ? (
          <Text color={rainbowColor(8)}>
            No quantum events yet. Run EXECUTE to evolve the graph.
          </Text>
        ) : (
          visualization.quantumEvents.map((event: string, index: number) => (
            <Text key={`event-${index}`} color={rainbowColor(index + 1)}>
              {event}
            </Text>
          ))
        )}
      </Box>
    </Box>
  );
});

const MetricsPanel = React.memo(function MetricsPanel({
  beta1,
  paths,
  activeWaveCount,
  phase,
}: {
  beta1: number;
  paths: number;
  activeWaveCount: number;
  phase: number;
}) {
  return (
    <Box
      flexDirection="row"
      justifyContent="space-between"
      paddingX={1}
      marginBottom={1}
    >
      <Box>
        <Text color={spectrumColor(3, phase)}>Betti Number (</Text>
        <Text color={spectrumColor(1, phase)} bold={!isDimBreath(phase)}>
          beta1
        </Text>
        <Text color={spectrumColor(3, phase)}>): </Text>
        <Text color={spectrumColor(4, phase)} bold={!isDimBreath(phase)}>
          {beta1}
        </Text>
      </Box>
      <Box>
        <Text color={spectrumColor(7, phase)}>Wave Amplitudes: </Text>
        <Text color={spectrumColor(9, phase)} bold={!isDimBreath(phase)}>
          {activeWaveCount > 0 ? activeWaveCount : paths}
        </Text>
      </Box>
    </Box>
  );
});

const SteeringPanel = React.memo(function SteeringPanel({
  steering,
  steeringMode,
  traceAlert,
  phase,
}: {
  steering: GnosisSteeringMetrics | null;
  steeringMode: GnosisSteeringMode;
  traceAlert: string | null;
  phase: number;
}) {
  if (!surfaceSteeringMetrics(steeringMode)) {
    return (
      <Box
        borderStyle="single"
        borderColor={rainbowColor(8)}
        paddingX={1}
        marginBottom={1}
      >
        <Text color={spectrumColor(8, phase)}>
          Steering disabled. Mode: {steeringMode}.
        </Text>
      </Box>
    );
  }

  if (!steering) {
    return (
      <Box
        borderStyle="single"
        borderColor={traceAlert ? 'redBright' : rainbowColor(8)}
        paddingX={1}
        marginBottom={1}
      >
        {traceAlert ? (
          <Text color="redBright" bold>
            {traceAlert}
          </Text>
        ) : null}
        <Text color={traceAlert ? 'redBright' : spectrumColor(8, phase)}>
          Steering idle. Draw topology to measure Wallace; execute to observe
          time in microCharleys.
        </Text>
      </Box>
    );
  }

  return (
    <Box
      flexDirection="column"
      borderStyle="single"
      borderColor={traceAlert ? 'redBright' : rainbowColor(9)}
      paddingX={1}
      marginBottom={1}
    >
      <Text bold color={rainbowColor(10)}>
        STEERING
      </Text>
      {traceAlert ? (
        <Text color="redBright" bold>
          {traceAlert}
        </Text>
      ) : null}
      <Text color={spectrumColor(7, phase)}>
        mode: {steering.mode} | Wallace metric: {steering.wallaceNumber} wallys
        | frontier-fill: {steering.frontierFill}
      </Text>
      <Text color={spectrumColor(4, phase)}>
        topology-deficit: {steering.topologyDeficit} | regime: {steering.regime}
        {' | '}bias: {steering.leanInBias} ({steering.boundaryWalkBias})
        {surfaceSteeringRecommendations(steering.mode) &&
        steering.recommendedAction
          ? ` | action: ${steering.recommendedAction}`
          : ''}
      </Text>
      {steering.failureBoundary.deterministicCollapseCandidate ||
      steering.failureBoundary.contagiousRepairPressure ? (
        <Text color={spectrumColor(5, phase)}>
          collapse cost: {steering.failureBoundary.collapseCostAction}
          {steering.failureBoundary.collapseCostFloor > 0
            ? ` | floor: ${steering.failureBoundary.collapseCostFloor} | paid: ${steering.failureBoundary.totalPaidCost}`
            : ''}
          {steering.failureBoundary.paidStageCount > 0
            ? ` | paid-stages: ${steering.failureBoundary.paidStageCount}`
            : ''}
          {steering.failureBoundary.zeroWasteCollapseRisk
            ? ' | free collapse blocked'
            : ''}
          {steering.failureBoundary.freeCollapsePrefixRisk
            ? ` | prefix deficit: ${steering.failureBoundary.prefixCostDeficit}`
            : ''}
          {steering.failureBoundary.contagiousRepairPressure
            ? ' | zero-vent repair pressure'
            : ''}
        </Text>
      ) : null}
      <Text color={spectrumColor(6, phase)}>
        frontier-median: {steering.eda.frontierWidths.summary.median.datum} |
        occupancy-median: {steering.eda.layerOccupancy.summary.median.datum} |
        graph-density: {steering.eda.graph.density} | outliers:{' '}
        {steering.eda.graphOutliers.length}
      </Text>
      {steering.telemetry.wallMicroCharleys !== null ||
      steering.telemetry.cpuMicroCharleys !== null ? (
        <Text color={spectrumColor(2, phase)}>
          wall time (uCharleys): {steering.telemetry.wallMicroCharleys ?? 'n/a'}{' '}
          | cpu time (uCharleys): {steering.telemetry.cpuMicroCharleys ?? 'n/a'}{' '}
          | wall/cpu: {steering.telemetry.wallToCpuRatio ?? 'n/a'}
        </Text>
      ) : null}
    </Box>
  );
});

const HistoryPanel = React.memo(function HistoryPanel({
  history,
}: {
  history: HistoryEntry[];
}) {
  return (
    <Box flexDirection="column" marginTop={1}>
      {history.map((entry: HistoryEntry, index: number) => {
        const markerColor =
          entry.type === 'input'
            ? rainbowColor(index + 2)
            : entry.type === 'error'
            ? 'redBright'
            : rainbowColor(index + 6);
        const textColor =
          entry.type === 'error' ? 'redBright' : rainbowColor(index + 1);

        return (
          <Box key={`history-${index}`} flexDirection="row">
            <Text color={markerColor}>
              {entry.type === 'input' ? '>> ' : '   '}
            </Text>
            <Text color={textColor}>{entry.text}</Text>
          </Box>
        );
      })}
    </Box>
  );
});

/**
 * Renders the start Repl view.
 */
export function startRepl(options: ReplOptions = {}) {
  const {
    verbose = false,
    steeringMode = DEFAULT_GNOSIS_STEERING_MODE,
    steeringTraceWindow = DEFAULT_GNOSIS_STEERING_TRACE_WINDOW,
    executionAuth = null,
    steeringRelayConfig = null,
  } = options;
  const betty = new BettyCompiler();
  const registry = new GnosisRegistry();
  const engine = new GnosisEngine(registry);
  const steeringTraceStore = surfaceSteeringMetrics(steeringMode)
    ? new GnosisSteeringTraceStore({
        recentWindow: steeringTraceWindow,
        relay: steeringRelayConfig,
        executionAuth,
      })
    : null;

  // Register basic handlers for REPL testing
  registry.register('Codec', async (payload, props) => {
    const type = props['type'] || 'unknown';
    return `[Codec:${type}] Encoded: ${payload}`;
  });

  const Repl = () => {
    const { exit } = useApp();
    const [phase, setPhase] = useState(0);
    const [history, setHistory] = useState<HistoryEntry[]>([]);
    const [query, setQuery] = useState('');
    const [sourceLines, setSourceLines] = useState<string[]>([]);
    const [ast, setAst] = useState<GraphAST | null>(null);
    const [metrics, setMetrics] = useState({ b1: 0, paths: 1 });
    const [steering, setSteering] = useState<GnosisSteeringMetrics | null>(
      null
    );
    const [analysisReport, setAnalysisReport] =
      useState<GnosisComplexityReport | null>(null);
    const [steeringTraceAlert, setSteeringTraceAlert] = useState<string | null>(
      null
    );
    const executionAuthRef = useRef<GnosisExecutionAuthContext | null>(
      executionAuth
    );
    const bootstrapExecutionAuthRef = useRef<GnosisExecutionAuthContext | null>(
      null
    );
    const [visualization, setVisualization] = useState<ReplVisualizationState>(
      createInitialVisualization(null)
    );

    const appendHistory = (...entries: HistoryEntry[]) => {
      setHistory((previousHistory: HistoryEntry[]) =>
        [...previousHistory, ...entries].slice(-MAX_HISTORY_ENTRIES)
      );
    };
    const visibleHistory = useMemo(
      () => history.slice(-MAX_DISPLAYED_HISTORY),
      [history]
    );

    useEffect(() => {
      const timer = setInterval(() => {
        setPhase((previousPhase: number) => previousPhase + 1);
      }, BREATH_INTERVAL_MS);

      return () => {
        clearInterval(timer);
      };
    }, []);

    useEffect(() => {
      if (!steeringTraceStore) {
        return undefined;
      }

      return () => {
        steeringTraceStore.disconnect();
      };
    }, []);

    const appendSteeringTraceHistory = async (
      report: GnosisComplexityReport,
      nextSteering: GnosisSteeringMetrics,
      outcome: 'ok' | 'error',
      errorMessage: string | null
    ): Promise<void> => {
      if (!steeringTraceStore) {
        return;
      }

      steeringTraceStore.setExecutionAuth(
        executionAuthRef.current ?? bootstrapExecutionAuthRef.current
      );
      await steeringTraceStore.connect();

      const traceRecord = createSteeringTraceRecord({
        command: 'repl',
        filePath: null,
        source: sourceLines.join('\n'),
        report,
        steering: nextSteering,
        payload: 'REPL_INIT',
        outcome,
        errorMessage,
      });
      steeringTraceStore.append(traceRecord);
      const steeringTraceSummary = steeringTraceStore.summarize({
        cohortKey: traceRecord.cohortKey,
      });
      const traceStatus = steeringTraceStore.status;
      const traceAlert = formatSteeringTraceAlert(
        steeringTraceSummary,
        traceStatus
      );
      setSteeringTraceAlert(traceAlert);

      appendHistory({
        type: outcome === 'error' ? 'error' : 'output',
        text: `[Steering Trace] ${formatSteeringTraceStoreStatus(
          traceStatus
        )} ${formatSteeringTraceSummary(steeringTraceSummary)}`,
      });

      if (traceAlert) {
        appendHistory({
          type: 'error',
          text: traceAlert,
        });
      }
    };

    const applyCompiledSource = async (
      nextSourceLines: string[],
      inputLabel: string
    ) => {
      if (nextSourceLines.length === 0) {
        bootstrapExecutionAuthRef.current = null;
        setSourceLines([]);
        setAst(null);
        setMetrics({ b1: 0, paths: 1 });
        setSteering(null);
        setAnalysisReport(null);
        setVisualization(createInitialVisualization(null));
        appendHistory(
          { type: 'input', text: inputLabel },
          { type: 'output', text: '[Editor] Topology cleared.' }
        );
        return;
      }

      try {
        const source = nextSourceLines.join('\n');
        const {
          ast: compiledAst,
          output,
          b1,
          diagnostics,
        } = betty.parse(source);
        const parsedAst = compiledAst ?? null;

        setSourceLines(nextSourceLines);
        setAst(parsedAst);
        setMetrics({ b1, paths: Math.max(1, b1 + 1) });
        setVisualization(createInitialVisualization(parsedAst));

        appendHistory(
          { type: 'input', text: inputLabel },
          {
            type: 'output',
            text: formatCompilationMessage(output, diagnostics),
          }
        );

        if (!surfaceSteeringMetrics(steeringMode)) {
          setSteering(null);
          setAnalysisReport(null);
          return;
        }

        if (!parsedAst) {
          bootstrapExecutionAuthRef.current = null;
          setSteering(null);
          setAnalysisReport(null);
          return;
        }

        try {
          const boundaryWalkExecutionAuth =
            await resolveBoundaryWalkExecutionAuth(
              parsedAst,
              executionAuthRef.current
            );
          bootstrapExecutionAuthRef.current = boundaryWalkExecutionAuth;
          assertBoundaryWalkAuthorized(
            source,
            steeringMode,
            boundaryWalkExecutionAuth
          );
          const report = await analyzeGnosisSource(source, { steeringMode });
          setAnalysisReport(report);
          setSteering(report.steering);
        } catch (error: unknown) {
          bootstrapExecutionAuthRef.current = null;
          setSteering(null);
          setAnalysisReport(null);
          appendHistory({
            type: 'error',
            text: `[Analysis Crash] ${formatMottoError(error)}`,
          });
        }
      } catch (error: unknown) {
        bootstrapExecutionAuthRef.current = null;
        setSteering(null);
        setAnalysisReport(null);
        appendHistory(
          { type: 'input', text: inputLabel },
          { type: 'error', text: `[Compiler Crash] ${formatMottoError(error)}` }
        );
      }
    };

    const handleSubmit = async (submittedQuery: string) => {
      const trimmedQuery = submittedQuery.trim();
      if (trimmedQuery.length === 0) {
        setQuery('');
        return;
      }

      const upperQuery = trimmedQuery.toUpperCase();
      if (upperQuery === 'EXIT') {
        exit();
        return;
      }

      if (upperQuery === 'CLEAR') {
        bootstrapExecutionAuthRef.current = null;
        setHistory([]);
        setSourceLines([]);
        setAst(null);
        setMetrics({ b1: 0, paths: 1 });
        setSteering(null);
        setAnalysisReport(null);
        setVisualization(createInitialVisualization(null));
        setQuery('');
        return;
      }

        if (upperQuery === 'SOURCE') {
          const sourceDump =
            sourceLines.length > 0
              ? sourceLines
                  .map((line: string, index: number) => `${index + 1}. ${line}`)
                  .join('\n')
              : '[Editor] Topology source is empty.';
        appendHistory(
          { type: 'input', text: 'SOURCE' },
          { type: 'output', text: sourceDump }
        );
        setQuery('');
        return;
      }

      if (upperQuery === 'UNDO') {
        if (sourceLines.length === 0) {
          appendHistory(
            { type: 'input', text: 'UNDO' },
            { type: 'error', text: `[Editor] ${ERROR_MOTTO}: Nothing to undo.` }
          );
        } else {
          const nextSourceLines = sourceLines.slice(0, -1);
          await applyCompiledSource(
            nextSourceLines,
            `UNDO (${sourceLines[sourceLines.length - 1]})`
          );
        }
        setQuery('');
        return;
      }

      if (upperQuery === 'EXECUTE') {
        if (!ast || ast.edges.length === 0) {
          appendHistory(
            { type: 'input', text: 'EXECUTE' },
            {
              type: 'error',
              text: `[Engine] ${ERROR_MOTTO}: No topology to execute. Add graph lines first.`,
            }
          );
          setQuery('');
          return;
        }

        try {
          const steeringStopwatch = startSteeringTelemetry();
          try {
            const executionOutput = await engine.execute(ast, 'REPL_INIT', {
              executionAuth: executionAuthRef.current,
            });
            executionAuthRef.current =
              engine.executionAuth ?? executionAuthRef.current;
            const runtimeTelemetry = finishSteeringTelemetry(steeringStopwatch);
            const waveState = buildVisualizationFromExecution(
              ast,
              executionOutput
            );
            const executedSteering =
              steering && analysisReport
                ? withSteeringTelemetry(steering, runtimeTelemetry)
                : null;
            setVisualization(waveState);
            setSteering(executedSteering);
            appendHistory(
              { type: 'input', text: 'EXECUTE' },
              { type: 'output', text: executionOutput }
            );
            if (executedSteering && analysisReport) {
              try {
                await appendSteeringTraceHistory(
                  analysisReport,
                  executedSteering,
                  'ok',
                  null
                );
              } catch (traceError: unknown) {
                const alert = `STEERING ALERT level=CRITICAL relay=disconnected room=${
                  steeringRelayConfig?.roomName ?? 'unknown'
                } relay-error="${formatMottoError(traceError)}"`;
                setSteeringTraceAlert(alert);
                appendHistory(
                  {
                    type: 'error',
                    text: `[STEERING TRACE RELAY FAILURE] ${formatMottoError(
                      traceError
                    )}`,
                  },
                  {
                    type: 'error',
                    text: alert,
                  }
                );
              }
            }
          } catch (error: unknown) {
            executionAuthRef.current =
              engine.executionAuth ?? executionAuthRef.current;
            const runtimeTelemetry = finishSteeringTelemetry(steeringStopwatch);
            const failedSteering =
              steering && analysisReport
                ? withSteeringTelemetry(steering, runtimeTelemetry)
                : null;
            appendHistory(
              { type: 'input', text: 'EXECUTE' },
              {
                type: 'error',
                text: `[Engine Crash] ${formatMottoError(error)}`,
              }
            );
            if (failedSteering && analysisReport) {
              setSteering(failedSteering);
              try {
                await appendSteeringTraceHistory(
                  analysisReport,
                  failedSteering,
                  'error',
                  formatError(error)
                );
              } catch (traceError: unknown) {
                const alert = `STEERING ALERT level=CRITICAL relay=disconnected room=${
                  steeringRelayConfig?.roomName ?? 'unknown'
                } relay-error="${formatMottoError(traceError)}"`;
                setSteeringTraceAlert(alert);
                appendHistory(
                  {
                    type: 'error',
                    text: `[STEERING TRACE RELAY FAILURE] ${formatMottoError(
                      traceError
                    )}`,
                  },
                  {
                    type: 'error',
                    text: alert,
                  }
                );
              }
            }
          }
        } catch (error: unknown) {
          appendHistory(
            { type: 'input', text: 'EXECUTE' },
            {
              type: 'error',
              text: `[Steering Crash] ${formatMottoError(error)}`,
            }
          );
        }

        setQuery('');
        return;
      }

      const nextSourceLines = [...sourceLines, submittedQuery];
      await applyCompiledSource(nextSourceLines, submittedQuery);
      setQuery('');
    };

    return (
      <Box flexDirection="column" padding={1} minHeight={20}>
        <SpectrumHeading phase={phase} verbose={verbose} />

        <MetricsPanel
          beta1={metrics.b1}
          paths={metrics.paths}
          activeWaveCount={visualization.activeWaveFunction.length}
          phase={phase}
        />
        <SteeringPanel
          steering={steering}
          steeringMode={steeringMode}
          traceAlert={steeringTraceAlert}
          phase={phase}
        />

        <SourcePanel sourceLines={sourceLines} />
        <GraphCanvas ast={ast} visualization={visualization} />
        <HistoryPanel history={visibleHistory} />

        <Box flexDirection="row" marginY={1}>
          <Text color={spectrumColor(0, phase)} bold={!isDimBreath(phase)}>
            {'>'}{' '}
          </Text>
          <TextInput
            value={query}
            onChange={setQuery}
            onSubmit={handleSubmit}
          />
        </Box>

        <Box marginY={1} flexDirection="row">
          <Text color={spectrumColor(3, phase)} dimColor={isDimBreath(phase)}>
            Commands:
          </Text>
          <Text color={spectrumColor(0, phase)}> EXECUTE</Text>
          <Text color={spectrumColor(2, phase)}> | SOURCE</Text>
          <Text color={spectrumColor(4, phase)}> | UNDO</Text>
          <Text color={spectrumColor(6, phase)}> | CLEAR</Text>
          <Text color={spectrumColor(8, phase)}> | EXIT</Text>
        </Box>
      </Box>
    );
  };

  render(<Repl />);
}
