// @a0n/gnosis/src/reynolds-bft.ts
export type FoldRegime = 'mergeAll' | 'quorumFold' | 'syncRequired';

function normalizeCount(value: number, name: string): number {
  if (!Number.isFinite(value)) {
    throw new Error(`${name} must be finite.`);
  }

  const normalized = Math.floor(value);
  if (normalized < 1) {
    throw new Error(`${name} must be >= 1.`);
  }

  return normalized;
}

/**
 * Handles the gnosis idle Stages workflow.
 */
export function idleStages(stages: number, chunks: number): number {
  const normalizedStages = normalizeCount(stages, 'stages');
  const normalizedChunks = normalizeCount(chunks, 'chunks');
  return normalizedStages - Math.min(normalizedStages, normalizedChunks);
}

/**
 * Handles the gnosis quorum Safe Fold workflow.
 */
export function quorumSafeFold(stages: number, chunks: number): boolean {
  const normalizedStages = normalizeCount(stages, 'stages');
  const idle = idleStages(normalizedStages, chunks);
  return 3 * idle < normalizedStages;
}

/**
 * Handles the gnosis majority Safe Fold workflow.
 */
export function majoritySafeFold(stages: number, chunks: number): boolean {
  const normalizedStages = normalizeCount(stages, 'stages');
  const idle = idleStages(normalizedStages, chunks);
  return 2 * idle < normalizedStages;
}

/**
 * Handles the gnosis classify Fold Regime workflow.
 */
export function classifyFoldRegime(stages: number, chunks: number): FoldRegime {
  const normalizedStages = normalizeCount(stages, 'stages');
  const normalizedChunks = normalizeCount(chunks, 'chunks');

  if (3 * normalizedChunks > 2 * normalizedStages) {
    return 'mergeAll';
  }

  if (2 * normalizedChunks > normalizedStages) {
    return 'quorumFold';
  }

  return 'syncRequired';
}
