// @a0n/gnosis/src/scheduling-mesh.ts
/**
 * Scheduling Mesh -- Contiguous Computation Surface
 *
 * A single mesh where every node is a vertex, every edge is a direct
 * connection, and fork/race/fold are vertex operations on the graph.
 * No message passing, no boundaries between schedulers -- one topology.
 *
 * The mesh is a k-gon tiling (triangle, square, or hexagon) that fills
 * space. Each vertex holds a god formula weight. Pathfinding replaces
 * message routing. Propagation replaces dispatch.
 *
 * Built on the three primitives:
 *   P1 (FORK): fan out from a vertex along k radials
 *   P2 (FOLD): collapse results at the hub
 *   P3 (RACE): first radial to complete wins
 *
 * @packageDocumentation
 */

import { godFormula } from './holy-trinity';
import { type MeshConfig, GNOSTIC_MESHES, type GnosticMeshLevel, resolveSyzygyDimension } from './spiderweb-topology';
import {
  scanCoreReductionFamilies,
  type CoreReductionScanResult,
} from './reduction-projection-scan';
import { buildReport, missingBackpressure } from './flow-obstruction';

// ─── Trit — The Fundamental Unit ────────────────────────────────────

/**
 * A trit: the three states of a scheduling vertex.
 *
 *   -1 (VENT):   compute/processing. The active surface of the Unscheduling Mesh.
 *    0 (VOID):   idle/vacuum. Exists but inactive. Ground state.
 *   +1 (FORK):   indexing/routing. The active surface of the Scheduling Mesh.
 *
 * In the Inverted Paradigm, the 1/3 FORK state operates strictly as an index
 * or routing address, while the 2/3 VENT/VOID space is leveraged to proactively
 * execute workloads in parallel.
 *
 * A trit carries log₂(3) ≈ 1.585 bits. More efficient per digit
 * than binary (3 is closer to e ≈ 2.718 than 2 is).
 */
export type Trit = -1 | 0 | 1;

export const VENT: Trit = -1;
export const VOID: Trit = 0;
export const FORK: Trit = 1;

/** Convert a trit to its name. */
export function tritName(t: Trit): string {
  return t === -1 ? 'vent' : t === 0 ? 'void' : 'fork';
}

/** A trit word: an array of trits. n trits = 3^n states. */
export type TritWord = Trit[];

/** Number of states in a trit word. */
export function tritStates(length: number): number {
  return 3 ** length;
}

// ─── Types ──────────────────────────────────────────────────────────

/** A vertex in the scheduling mesh. */
export interface MeshVertex {
  /** Unique vertex ID */
  readonly id: number;
  /** Vertex type: hub (coordinator), relay (intermediate), endpoint (edge) */
  readonly type: 'hub' | 'relay' | 'endpoint';
  /** Position in the mesh (cell index, ring, radial) */
  readonly cell: number;
  readonly ring: number;
  readonly radial: number;
  /** God formula weight: w = R - min(v, R) + 1 */
  weight: number;
  /** Rejection count on this vertex */
  rejections: number;
  /** Current payload (task being processed) */
  payload: unknown | null;
  /** Vertex state as a trit: -1 (vented), 0 (idle), +1 (active) */
  state: Trit;
  /** Whether this vertex is currently busy (set by MeshEngine) */
  busy: boolean;
}

/** An edge in the scheduling mesh. */
export interface MeshEdge {
  /** Source vertex ID */
  readonly from: number;
  /** Target vertex ID */
  readonly to: number;
  /** Edge type: radial (fast lane), spiral (cross-link), belt (frame), bond (inter-cell) */
  readonly type: 'radial' | 'spiral' | 'belt' | 'bond';
  /** Traversal cost (1 for all edges -- uniform mesh) */
  readonly cost: number;
}

/** Result of a fork operation. */
export interface ForkResult {
  /** The hub vertex that initiated the fork */
  readonly hub: number;
  /** The radial vertices that received the forked task */
  readonly targets: number[];
  /** The beta1 of this fork (number of independent paths) */
  readonly beta1: number;
}

/** Result of a race operation. */
export interface RaceResult<T> {
  /** The winning vertex ID */
  readonly winner: number;
  /** The result from the winner */
  readonly result: T;
  /** Time taken (hops traversed) */
  readonly hops: number;
  /** Vertices that lost the race */
  readonly losers: number[];
}

/** Result of a fold operation. */
export interface FoldResult<T> {
  /** The hub vertex that collected results */
  readonly hub: number;
  /** All collected results */
  readonly results: T[];
  /** Number of results folded */
  readonly count: number;
}

/** A mesh event -- a neuron firing. */
export type MeshEvent =
  | { type: 'fork'; hub: number; targets: number[]; hop: number }
  | { type: 'race'; winner: number; losers: number[]; hop: number }
  | { type: 'fold'; hub: number; hop: number }
  | { type: 'propagate'; from: number; to: number; hops: number; hop: number }
  | { type: 'saturated'; hop: number }
  | { type: 'delayedFork'; hub: number; hop: number; depth: number }
  | { type: 'projection'; depth: number; tritStates: number; bitStates: number };

/** A firing record in the activation trace. */
export interface Firing {
  hub: number;
  type: 'fork' | 'race' | 'fold' | 'propagate';
  hop: number;
  winner?: number;
  target?: number;
}

/** The complete trace of an activation cascade. */
export interface ActivationTrace<T> {
  firings: Firing[];
  results: T[];
  totalHops: number;
  cellsActivated: number;
  startHub: number;
}

// ─── The Mesh ───────────────────────────────────────────────────────

/**
 * SchedulingMesh -- a contiguous computation surface.
 *
 * One graph. No boundaries. Fork/race/fold are vertex operations.
 * Pathfinding replaces message routing.
 */
export class SchedulingMesh {
  /** All vertices in the mesh */
  readonly vertices: MeshVertex[];
  /** All edges in the mesh (adjacency list) */
  readonly adjacency: Map<number, MeshEdge[]>;
  /** Cell count (number of k-gon cells) */
  readonly cellCount: number;
  /** The mesh configuration (k peers, s hops) */
  readonly config: MeshConfig;
  /** Hub vertices (one per cell) */
  readonly hubs: number[];
  /** Total rejection rounds for god formula */
  private rounds: number;

  constructor(level: GnosticMeshLevel, cells: number) {
    this.config = GNOSTIC_MESHES[level];
    this.cellCount = cells;
    this.rounds = 0;
    this.vertices = [];
    this.adjacency = new Map();
    this.hubs = [];

    this.buildMesh();
  }

  /** Number of vertices in the mesh. */
  get size(): number {
    return this.vertices.length;
  }

  /** Beta1 per cell. */
  get beta1(): number {
    return this.config.peers * this.config.hops;
  }

  /** Total beta1 across all cells. */
  get totalBeta1(): number {
    return this.beta1 * this.cellCount;
  }

  /** Composite dimension: (cells - 1) * beta1 + 1. */
  get compositeDimension(): number {
    if (this.cellCount <= 0) return 0;
    return (this.cellCount - 1) * this.beta1 + 1;
  }

  // ─── Construction ───────────────────────────────────────────────

  private buildMesh(): void {
    const { peers: k, hops: s } = this.config;
    let vertexId = 0;

    for (let cell = 0; cell < this.cellCount; cell++) {
      const cellOffset = vertexId;

      // Hub vertex
      const hubId = vertexId++;
      this.vertices.push({
        id: hubId, type: 'hub',
        cell, ring: -1, radial: -1,
        weight: 1, rejections: 0,
        payload: null, state: VOID, busy: false,
      });
      this.hubs.push(hubId);
      this.adjacency.set(hubId, []);

      // Relay vertices: k radials × s rings
      for (let r = 0; r < k; r++) {
        for (let ring = 0; ring < s; ring++) {
          const relayId = vertexId++;
          this.vertices.push({
            id: relayId, type: 'relay',
            cell, ring, radial: r,
            weight: 1, rejections: 0,
            payload: null, state: VOID, busy: false,
          });
          this.adjacency.set(relayId, []);
        }
      }

      // Endpoint vertices: k endpoints
      for (let r = 0; r < k; r++) {
        const epId = vertexId++;
        this.vertices.push({
          id: epId, type: 'endpoint',
          cell, ring: s, radial: r,
          weight: 1, rejections: 0,
          payload: null, state: VOID, busy: false,
        });
        this.adjacency.set(epId, []);
      }

      // ─── Edges within cell ──────────────────────────────────

      const relayIndex = (radial: number, ring: number) =>
        cellOffset + 1 + radial * s + ring;
      const epIndex = (radial: number) =>
        cellOffset + 1 + k * s + radial;

      // Radial edges: hub → first relay → ... → last relay → endpoint
      for (let r = 0; r < k; r++) {
        if (s > 0) {
          this.addEdge(hubId, relayIndex(r, 0), 'radial');
          for (let ring = 1; ring < s; ring++) {
            this.addEdge(relayIndex(r, ring - 1), relayIndex(r, ring), 'radial');
          }
          this.addEdge(relayIndex(r, s - 1), epIndex(r), 'radial');
        } else {
          this.addEdge(hubId, epIndex(r), 'radial');
        }
      }

      // Spiral edges: connect adjacent radials within each ring
      for (let ring = 0; ring < s; ring++) {
        for (let r = 0; r < k; r++) {
          this.addEdge(relayIndex(r, ring), relayIndex((r + 1) % k, ring), 'spiral');
        }
      }

      // Belt edges: connect adjacent endpoints
      for (let r = 0; r < k; r++) {
        this.addEdge(epIndex(r), epIndex((r + 1) % k), 'belt');
      }
    }

    // ─── Inter-cell bonds ───────────────────────────────────────
    // Connect adjacent cells via their closest endpoints
    for (let i = 0; i < this.cellCount - 1; i++) {
      this.addCellBond(i, i + 1);
    }
    // Ring closure for > 2 cells
    if (this.cellCount > 2) {
      this.addCellBond(this.cellCount - 1, 0);
    }
  }

  private addEdge(from: number, to: number, type: MeshEdge['type']): void {
    const edge: MeshEdge = { from, to, type, cost: 1 };
    const reverse: MeshEdge = { from: to, to: from, type, cost: 1 };
    this.adjacency.get(from)!.push(edge);
    this.adjacency.get(to)!.push(reverse);
  }

  private addCellBond(cellA: number, cellB: number): void {
    // Bond the first endpoint of cellA to the first endpoint of cellB
    const { peers: k, hops: s } = this.config;
    const verticesPerCell = k * s + k + 1;
    const epA = cellA * verticesPerCell + 1 + k * s; // first endpoint of cellA
    const epB = cellB * verticesPerCell + 1 + k * s; // first endpoint of cellB
    this.addEdge(epA, epB, 'bond');
  }

  // ─── Fork / Race / Fold ─────────────────────────────────────────

  /**
   * P1 FORK: Fan out a task from a hub vertex along all k radials.
   * Each radial's first relay (or endpoint if s=0) receives the task.
   * Returns the fork targets.
   */
  fork(hubId: number, task: unknown): ForkResult {
    const hub = this.vertices[hubId];
    if (!hub || hub.type !== 'hub') {
      throw new Error(`Vertex ${hubId} is not a hub`);
    }

    const edges = this.adjacency.get(hubId) ?? [];
    const radialEdges = edges.filter((e: any) => e.type === 'radial');
    const targets: number[] = [];

    for (const edge of radialEdges) {
      const target = this.vertices[edge.to]!;
      target.payload = task;
      target.state = FORK;
      targets.push(edge.to);
    }

    hub.state = FORK;
    this.rounds++;

    return {
      hub: hubId,
      targets,
      beta1: this.beta1,
    };
  }

  /**
   * P3 RACE / PROACTIVE VENT: Execute the workload across all targets.
   * In the Inverted Paradigm, the targets (radials) proactively enter the VENT
   * state to process the workload. The hub (FORK) acts merely as the index.
   * All paths do the work. The "winner" is simply the routing path selected 
   * by the god formula, but the compute volume used the full 2/3 negative space.
   */
  race<T>(targets: number[], evaluate: (vertexId: number) => T): RaceResult<T> {
    if (targets.length === 0) {
      throw new Error('Cannot race with no targets');
    }

    // INVERTED PARADIGM: Proactively push all targets into the compute surface
    let winnerId = targets[0]!;
    let bestWeight = this.vertices[winnerId]!.weight;
    
    // Cache the proactively computed results because we don't want to double-execute
    const results = new Map<number, T>();

    for (const id of targets) {
      const v = this.vertices[id]!;
      v.state = VENT; // -1: active compute surface
      
      // Apply the Bizarro Operator: evaluate ALL targets in the void's negative space.
      // They all do work. We ignore the return value of losers for now, 
      // but the machine cycles their compute.
      results.set(id, evaluate(id));

      if (v.weight > bestWeight) {
        bestWeight = v.weight;
        winnerId = id;
      }
    }

    // Retrieve the definitive result from the selected routing winner
    const result = results.get(winnerId)!;
    const losers: number[] = [];

    for (const id of targets) {
      if (id === winnerId) {
        this.vertices[id]!.state = VOID; // Winner returns to idle
      } else {
        // Losers completed their compute proactively. They increment their
        // rejections to adjust the routing topography for the next epoch.
        this.vertices[id]!.rejections++;
        this.vertices[id]!.weight = godFormula(this.rounds, this.vertices[id]!.rejections);
        this.vertices[id]!.state = VOID; // Return to idle
        this.vertices[id]!.payload = null;
        losers.push(id);
      }
    }

    const winner = this.vertices[winnerId]!;
    const hops = winner.ring >= 0 ? winner.ring + 1 : 1;

    return { winner: winnerId, result, hops, losers };
  }

  /**
   * P2 FOLD: Collect all results from target vertices back to the hub.
   * Resets all targets to idle.
   */
  fold<T>(hubId: number, targets: number[], extract: (vertexId: number) => T): FoldResult<T> {
    const results: T[] = [];

    for (const id of targets) {
      const v = this.vertices[id]!;
      if (v.payload !== null) {
        results.push(extract(id));
      }
      v.state = VOID;
      v.payload = null;
    }

    this.vertices[hubId]!.state = VOID;

    return { hub: hubId, results, count: results.length };
  }

  // ─── Pathfinding ────────────────────────────────────────────────

  /**
   * BFS shortest path between two vertices.
   * Returns the path as vertex IDs, or null if unreachable.
   */
  shortestPath(from: number, to: number): number[] | null {
    if (from === to) return [from];

    const visited = new Set<number>();
    const parent = new Map<number, number>();
    const queue: number[] = [from];
    visited.add(from);

    while (queue.length > 0) {
      const current = queue.shift()!;
      const edges = this.adjacency.get(current) ?? [];

      for (const edge of edges) {
        if (!visited.has(edge.to)) {
          visited.add(edge.to);
          parent.set(edge.to, current);

          if (edge.to === to) {
            // Reconstruct path
            const path: number[] = [to];
            let node = to;
            while (parent.has(node)) {
              node = parent.get(node)!;
              path.unshift(node);
            }
            return path;
          }

          queue.push(edge.to);
        }
      }
    }

    return null; // Unreachable
  }

  /**
   * Propagate a task through the mesh from a source vertex.
   * Visits all reachable vertices in BFS order, applying the visitor.
   * Returns the number of vertices reached.
   */
  propagate(source: number, visitor: (vertex: MeshVertex, hop: number) => boolean): number {
    const visited = new Set<number>();
    const queue: Array<{ id: number; hop: number }> = [{ id: source, hop: 0 }];
    visited.add(source);
    let reached = 0;

    while (queue.length > 0) {
      const { id, hop } = queue.shift()!;
      const vertex = this.vertices[id]!;

      const shouldContinue = visitor(vertex, hop);
      reached++;

      if (!shouldContinue) continue;

      const edges = this.adjacency.get(id) ?? [];
      for (const edge of edges) {
        if (!visited.has(edge.to)) {
          visited.add(edge.to);
          queue.push({ id: edge.to, hop: hop + 1 });
        }
      }
    }

    return reached;
  }

  // ─── Cascade — The Brain Mesh ─────────────────────────────────

  /** Listeners for mesh events */
  private listeners: Array<(event: MeshEvent) => void> = [];

  /** Subscribe to mesh events (neuron firings). */
  on(listener: (event: MeshEvent) => void): void {
    this.listeners.push(listener);
  }

  private emit(event: MeshEvent): void {
    for (const fn of this.listeners) fn(event);
  }

  /**
   * Activate a hub -- the cascade begins.
   *
   * 1. FORK: fan out the task along k radials
   * 2. RACE: each radial processes, highest-weight wins
   * 3. FOLD: collect the winner's result at the hub
   * 4. PATHFIND: find the nearest idle hub
   * 5. PROPAGATE: send the result to that hub
   * 6. REPEAT: the next hub activates
   *
   * The cascade stops when:
   * - No idle hubs remain (mesh saturated)
   * - maxHops exceeded (depth limit)
   * - The task function returns null (task complete)
   *
   * This formalizes the perpetual scheduling machine.
   */
  activate<T>(
    startHub: number,
    task: unknown,
    process: (vertexId: number, payload: unknown) => T | null,
    maxHops: number = this.cellCount,
  ): ActivationTrace<T> {
    const trace: ActivationTrace<T> = {
      firings: [],
      results: [],
      totalHops: 0,
      cellsActivated: 0,
      startHub,
    };

    let currentHub = startHub;
    let currentPayload: unknown = task;

    for (let hop = 0; hop < maxHops; hop++) {
      // 1. FORK
      const forkResult = this.fork(currentHub, currentPayload);
      this.emit({ type: 'fork', hub: currentHub, targets: forkResult.targets, hop });
      trace.firings.push({ hub: currentHub, type: 'fork', hop });

      // 2. RACE
      const raceResult = this.race<T | null>(forkResult.targets, (vid) => {
        const vertex = this.vertices[vid];
        if (vertex === undefined) {
          throw new Error(`Missing vertex ${vid} during race`);
        }
        return process(vid, vertex.payload);
      });
      this.emit({ type: 'race', winner: raceResult.winner, losers: raceResult.losers, hop });
      trace.firings.push({ hub: currentHub, type: 'race', hop, winner: raceResult.winner });

      // Task signaled completion
      if (raceResult.result === null) {
        this.fold(currentHub, forkResult.targets, () => null);
        this.emit({ type: 'fold', hub: currentHub, hop });
        trace.firings.push({ hub: currentHub, type: 'fold', hop });
        break;
      }

      trace.results.push(raceResult.result);

      // 3. FOLD
      this.fold(currentHub, forkResult.targets, () => null);
      this.emit({ type: 'fold', hub: currentHub, hop });
      trace.firings.push({ hub: currentHub, type: 'fold', hop });
      trace.cellsActivated++;

      // 4. PATHFIND to nearest idle hub
      const idleTargets = this.idleHubs();
      if (idleTargets.length === 0) {
        this.emit({ type: 'saturated', hop });
        break;
      }

      // Find closest idle hub by BFS
      let nearestHub: number | null = null;
      let nearestDist = Infinity;
      const winnerVertex = raceResult.winner;

      for (const candidateHub of idleTargets) {
        const path = this.shortestPath(winnerVertex, candidateHub);
        if (path && path.length < nearestDist) {
          nearestDist = path.length;
          nearestHub = candidateHub;
        }
      }

      if (nearestHub === null) break;

      // 5. PROPAGATE
      this.emit({ type: 'propagate', from: winnerVertex, to: nearestHub, hops: nearestDist, hop });
      trace.firings.push({ hub: currentHub, type: 'propagate', hop, target: nearestHub });
      trace.totalHops += nearestDist;

      // 6. REPEAT -- the cascade continues
      currentHub = nearestHub;
      currentPayload = raceResult.result;
    }

    return trace;
  }

  /**
   * Fire all hubs simultaneously -- full mesh activation.
   * Each hub gets its own task. The mesh processes everything in parallel.
   */
  fireAll<T>(
    taskForCell: (cell: number) => unknown,
    process: (vertexId: number, payload: unknown) => T | null,
  ): T[] {
    const results: T[] = [];

    for (let cell = 0; cell < this.cellCount; cell++) {
      const hubId = this.hubs[cell]!;
      if (this.vertices[hubId]!.state !== VOID) continue;

      const task = taskForCell(cell);
      const forkResult = this.fork(hubId, task);

      const raceResult = this.race<T | null>(forkResult.targets, (vid) => {
        return process(vid, this.vertices[vid]!.payload);
      });

      if (raceResult.result !== null) {
        results.push(raceResult.result);
      }

      this.fold(hubId, forkResult.targets, () => null);
    }

    return results;
  }

  /**
   * Reset all vertices to idle with weight 1.
   */
  reset(): void {
    for (const v of this.vertices) {
      v.weight = 1;
      v.rejections = 0;
      v.payload = null;
      v.state = VOID;
    }
    this.rounds = 0;
  }

  // ─── Queries ────────────────────────────────────────────────────

  /** Get all idle (non-busy) hub vertices. */
  idleHubs(): number[] {
    return this.hubs.filter((id: any) => this.vertices[id]!.state === VOID);
  }

  /** Get all neighbors of a vertex. */
  neighbors(vertexId: number): number[] {
    return (this.adjacency.get(vertexId) ?? []).map(e => e.to);
  }

  /** Get the hub of a given cell. */
  hubOfCell(cell: number): number {
    return this.hubs[cell]!;
  }

  /** Get all vertices in a given cell. */
  cellVertices(cell: number): MeshVertex[] {
    return this.vertices.filter(v => v.cell === cell);
  }

  /** Mesh statistics. */
  stats(): {
    vertices: number;
    edges: number;
    cells: number;
    beta1PerCell: number;
    totalBeta1: number;
    compositeDim: number;
    activeVertices: number;
    ventedVertices: number;
    totalRejections: number;
  } {
    let edgeCount = 0;
    for (const edges of this.adjacency.values()) {
      edgeCount += edges.length;
    }
    edgeCount /= 2; // Each edge counted twice (bidirectional)

    return {
      vertices: this.size,
      edges: edgeCount,
      cells: this.cellCount,
      beta1PerCell: this.beta1,
      totalBeta1: this.totalBeta1,
      compositeDim: this.compositeDimension,
      activeVertices: this.vertices.filter(v => v.state === FORK).length,
      ventedVertices: this.vertices.filter(v => v.state === VENT).length,
      totalRejections: this.vertices.reduce((sum, v) => sum + v.rejections, 0),
    };
  }
}

// ─── Factory ──────────────────────────────────────────────────────

/**
 * Create a scheduling mesh at a gnostic level with n cells.
 *
 * @param level - Gnostic mesh level (bythos through pleroma)
 * @param cells - Number of cells in the mesh
 * @returns A contiguous scheduling mesh
 */
export function createMesh(level: GnosticMeshLevel, cells: number): SchedulingMesh {
  return new SchedulingMesh(level, cells);
}

/**
 * Create a scheduling mesh auto-scaled to a parameter count.
 */
export function createMeshForParameters(numParameters: number, maxCells: number = 16): SchedulingMesh {
  const level = resolveSyzygyDimension(numParameters) as GnosticMeshLevel;
  const cells = Math.min(maxCells, Math.max(1, Math.ceil(Math.log2(numParameters / 100000))));
  return new SchedulingMesh(level, cells);
}

// ═══════════════════════════════════════════════════════════════════════
// BrainMesh -- Brains are nodes. The mesh of meshes.
// ═══════════════════════════════════════════════════════════════════════

/**
 * A brain -- a SchedulingMesh that is itself a node in a larger mesh.
 * Each brain has its own internal topology (k-gon cells, fork/race/fold),
 * AND connections to other brains. A brain can activate internally
 * or propagate to neighboring brains.
 *
 * The composition law guarantees: n brains at dimension d compose
 * to dimension (n-1)*d + 1. Brains composing brains composing brains.
 */
export interface Brain {
  /** Unique brain ID */
  readonly id: number;
  /** The internal scheduling mesh */
  readonly mesh: SchedulingMesh;
  /** God formula weight of this brain as a node */
  weight: number;
  /** Rejection count of this brain as a node */
  rejections: number;
  /** Whether this brain is currently processing */
  busy: boolean;
  /** State of the brain (VOID, FORK, VENT) */
  state?: number;
}

/** An event in the brain mesh. */
export type BrainEvent =
  | { type: 'brain-activate'; brainId: number; hop: number }
  | { type: 'brain-propagate'; from: number; to: number; hop: number }
  | { type: 'brain-complete'; brainId: number; hop: number }
  | { type: 'brain-saturated'; hop: number };

/** Trace of a brain mesh activation. */
export interface BrainTrace<T> {
  brainsActivated: number;
  results: T[];
  totalPropagations: number;
  firings: BrainEvent[];
}

export interface BrainMeshProjectionStats {
  readonly scan: CoreReductionScanResult;
  readonly totalRejections: number;
  readonly topologyEdges: number;
}

/**
 * BrainMesh -- a mesh of meshes.
 *
 * Each node is a full SchedulingMesh (a brain). The brains are connected
 * in their own topology. When a brain completes a task, it propagates
 * the result to the nearest idle brain along the brain-level mesh.
 *
 * This is the second level of the composition tower:
 *   Level 0: vertices in a mesh (fork/race/fold)
 *   Level 1: meshes in a brain mesh (activate/propagate/cascade)
 *   Level 2: brain meshes in a... (turtles all the way down)
 *
 * The god formula operates at every level. Brains that produce
 * results keep high weight. Brains that fail get rejected.
 * The brain mesh self-organizes just like its internal meshes.
 */
export class BrainMesh {
  readonly brains: Brain[];
  readonly adjacency: Map<number, number[]>;
  private rounds: number = 0;
  private listeners: Array<(event: BrainEvent) => void> = [];

  /**
   * @param brainLevel - Gnostic level for each brain's internal mesh
   * @param cellsPerBrain - Cells in each brain's internal mesh
   * @param brainCount - Number of brains in the brain mesh
   * @param brainTopology - How brains connect ('ring' | 'chain' | 'complete' | 'grid')
   */
  constructor(
    brainLevel: GnosticMeshLevel,
    cellsPerBrain: number,
    brainCount: number,
    brainTopology: 'ring' | 'chain' | 'complete' | 'grid' = 'ring',
  ) {
    this.brains = [];
    this.adjacency = new Map();

    // Create brains
    for (let i = 0; i < brainCount; i++) {
      this.brains.push({
        id: i,
        mesh: new SchedulingMesh(brainLevel, cellsPerBrain),
        weight: 1,
        rejections: 0,
        busy: false,
        state: VOID,
      });
      this.adjacency.set(i, []);
    }

    // Connect brains
    this.connectBrains(brainTopology, brainCount);
  }

  private connectBrains(topology: string, n: number): void {
    if (topology === 'chain') {
      for (let i = 0; i < n - 1; i++) this.addBrainEdge(i, i + 1);
    } else if (topology === 'complete') {
      for (let i = 0; i < n; i++)
        for (let j = i + 1; j < n; j++) this.addBrainEdge(i, j);
    } else if (topology === 'grid') {
      const cols = Math.ceil(Math.sqrt(n));
      for (let i = 0; i < n; i++) {
        if (i + 1 < n && (i + 1) % cols !== 0) this.addBrainEdge(i, i + 1);
        if (i + cols < n) this.addBrainEdge(i, i + cols);
      }
    } else {
      // ring
      for (let i = 0; i < n; i++) this.addBrainEdge(i, (i + 1) % n);
    }
  }

  private addBrainEdge(a: number, b: number): void {
    this.adjacency.get(a)!.push(b);
    this.adjacency.get(b)!.push(a);
  }

  on(listener: (event: BrainEvent) => void): void {
    this.listeners.push(listener);
  }

  private emit(event: BrainEvent): void {
    for (const fn of this.listeners) fn(event);
  }

  /** Get idle brains. */
  idleBrains(): number[] {
    return this.brains.filter(b => !b.busy).map(b => b.id);
  }

  /** Find nearest idle brain from a given brain via BFS. */
  nearestIdleBrain(from: number): number | null {
    const visited = new Set<number>();
    const queue = [from];
    visited.add(from);

    while (queue.length > 0) {
      const current = queue.shift()!;
      const neighbors = this.adjacency.get(current) ?? [];
      for (const n of neighbors) {
        if (!visited.has(n)) {
          visited.add(n);
          if (!this.brains[n]!.busy) return n;
          queue.push(n);
        }
      }
    }
    return null;
  }

  /**
   * Activate the brain mesh -- cascade across brains.
   *
   * 1. Activate a brain internally (full fork/race/fold cascade)
   * 2. Brain produces a result
   * 3. Pathfind to nearest idle brain
   * 4. Propagate result to that brain
   * 5. That brain activates internally
   * 6. Repeat
   *
   * Each brain is a universe of computation. The brain mesh
   * routes between universes.
   */
  activate<T>(
    startBrain: number,
    task: unknown,
    process: (vertexId: number, payload: unknown, brainId: number) => T | null,
    maxBrains: number = this.brains.length,
  ): BrainTrace<T> {
    const trace: BrainTrace<T> = {
      brainsActivated: 0,
      results: [],
      totalPropagations: 0,
      firings: [],
    };

    let currentBrain = startBrain;
    let currentPayload: unknown = task;

    for (let hop = 0; hop < maxBrains; hop++) {
      const brain = this.brains[currentBrain]!;
      brain.state = FORK;
      this.rounds++;

      this.emit({ type: 'brain-activate', brainId: currentBrain, hop });
      trace.firings.push({ type: 'brain-activate', brainId: currentBrain, hop });

      // Internal cascade within this brain
      const startHub = brain.mesh.idleHubs()[0] ?? brain.mesh.hubs[0]!;
      const internalTrace = brain.mesh.activate<T>(
        startHub,
        currentPayload,
        (vid, payload) => process(vid, payload, currentBrain),
      );

      const lastResult = internalTrace.results[internalTrace.results.length - 1] ?? null;

      brain.state = VOID;
      trace.brainsActivated++;

      if (lastResult === null) {
        this.emit({ type: 'brain-complete', brainId: currentBrain, hop });
        trace.firings.push({ type: 'brain-complete', brainId: currentBrain, hop });
        break;
      }

      trace.results.push(lastResult);

      // Pathfind to next idle brain
      const nextBrain = this.nearestIdleBrain(currentBrain);
      if (nextBrain === null) {
        this.emit({ type: 'brain-saturated', hop });
        trace.firings.push({ type: 'brain-saturated', hop });
        break;
      }

      // Reject losing brains (neighbors that weren't picked)
      const neighbors = this.adjacency.get(currentBrain) ?? [];
      for (const n of neighbors) {
        if (n !== nextBrain) {
          this.brains[n]!.rejections++;
          this.brains[n]!.weight = godFormula(this.rounds, this.brains[n]!.rejections);
        }
      }

      this.emit({ type: 'brain-propagate', from: currentBrain, to: nextBrain, hop });
      trace.firings.push({ type: 'brain-propagate', from: currentBrain, to: nextBrain, hop });
      trace.totalPropagations++;

      currentBrain = nextBrain;
      currentPayload = lastResult;
    }

    return trace;
  }

  /** Composite dimension of the brain mesh. */
  get compositeDimension(): number {
    if (this.brains.length === 0) return 0;
    const brainDim = this.brains[0]!.mesh.compositeDimension;
    return (this.brains.length - 1) * brainDim + 1;
  }

  /** Reset all brains and their internal meshes. */
  reset(): void {
    for (const brain of this.brains) {
      brain.weight = 1;
      brain.rejections = 0;
      brain.state = VOID;
      brain.mesh.reset();
    }
    this.rounds = 0;
  }

  /** Stats for the entire brain mesh. */
  stats(): {
    brains: number;
    verticesPerBrain: number;
    totalVertices: number;
    compositeDim: number;
    brainCompositeDim: number;
    projection: BrainMeshProjectionStats;
  } {
    const perBrain = this.brains[0]?.mesh.size ?? 0;
    const projection = this.projectionStats();
    return {
      brains: this.brains.length,
      verticesPerBrain: perBrain,
      totalVertices: perBrain * this.brains.length,
      compositeDim: this.brains[0]?.mesh.compositeDimension ?? 0,
      brainCompositeDim: this.compositeDimension,
      projection,
    };
  }

  /** Cross-reduction projection scan derived from current brain-mesh state. */
  projectionStats(): BrainMeshProjectionStats {
    const totalRejections = this.brains.reduce(
      (sum, brain) => sum + Math.max(0, Math.trunc(brain.rejections)),
      0
    );
    const topologyEdges = Math.floor(
      Array.from(this.adjacency.values()).reduce(
        (sum, neighbors) => sum + neighbors.length,
        0
      ) / 2
    );

    const primaryMesh = this.brains[0]?.mesh;
    const meshConfig = primaryMesh?.config ?? GNOSTIC_MESHES.proton;
    const obstructionReport =
      totalRejections > 0
        ? buildReport(
            'brain-mesh-runtime',
            [missingBackpressure(totalRejections)],
            Math.max(1, this.brains.length)
          )
        : buildReport('brain-mesh-runtime', [], Math.max(1, this.brains.length));

    const scan = scanCoreReductionFamilies({
      voidSignals: [
        {
          id: 'brain:void',
          dimension: 'cross_file',
          visibleCrossings: this.brains.length > 1 ? 1 : 0,
          intermoduleEdges: topologyEdges,
          exactWitness: false,
        },
      ],
      untanglingSignals: [
        {
          id: 'brain:untangling',
          crossingNumber: totalRejections + Math.max(0, this.brains.length - 1),
          beta1: primaryMesh?.beta1 ?? 0,
          violationPaths: Math.max(0, this.brains.length - 1),
          defectFamily: this.brains.length >= 7 ? 'composite' : null,
        },
      ],
      contagionSignals: [
        {
          id: 'brain:contagion',
          mesh: meshConfig,
          vector: {
            payloadId: 'brain-cascade',
            momentum: Math.max(1, meshConfig.hops),
            relayFriction: 1,
          },
          observedDim: 2,
        },
      ],
      obstructionSignals: [
        {
          id: 'brain:obstruction',
          report: obstructionReport,
          observedDimFloor: Math.max(1, this.brains.length),
        },
      ],
    });

    return {
      scan,
      totalRejections,
      topologyEdges,
    };
  }
}

/**
 * Create a brain mesh.
 */
export function createBrainMesh(
  brainLevel: GnosticMeshLevel = 'proton',
  cellsPerBrain: number = 3,
  brainCount: number = 7,
  topology: 'ring' | 'chain' | 'complete' | 'grid' = 'ring',
): BrainMesh {
  return new BrainMesh(brainLevel, cellsPerBrain, brainCount, topology);
}

// ═══════════════════════════════════════════════════════════════════════
// UniverseMesh -- The composition tower. Turtles all the way up.
// ═══════════════════════════════════════════════════════════════════════

/**
 * A layer in the composition tower.
 * Each layer contains nodes. Each node is itself a layer (or a leaf mesh).
 */
export interface TowerNode {
  readonly id: number;
  readonly depth: number;
  weight: number;
  rejections: number;
  busy: boolean;
  /** If leaf: the scheduling mesh. If branch: child tower. */
  readonly children: TowerNode[] | SchedulingMesh;
  /** Neighbor IDs at this level */
  readonly neighbors: number[];
  /** Node state */
  state?: number;
}

/**
 * UniverseMesh -- the recursive composition tower.
 *
 * Level 0: SchedulingMesh (vertices, fork/race/fold)
 * Level 1: array of Level 0 meshes connected in a topology
 * Level 2: array of Level 1 arrays connected in a topology
 * ...
 * Level N: one node containing the entire universe
 *
 * At every level:
 *   - fork fans out to children
 *   - race selects the highest-weight child
 *   - fold collects results
 *   - god formula governs weights
 *   - composition law computes dimension
 *
 * The tower has no fixed height. You build as deep as you need.
 * Each level multiplies dimension by (n-1)*d + 1.
 */
export class UniverseMesh {
  readonly root: TowerNode[];
  readonly depth: number;
  readonly topology: 'ring' | 'chain' | 'complete' | 'grid';
  private rounds: number = 0;
  private listeners: Array<(depth: number, type: string, nodeId: number) => void> = [];

  /**
   * Build a universe mesh.
   *
   * @param depth - Number of levels in the tower
   * @param nodesPerLevel - How many nodes at each level
   * @param leafLevel - Gnostic level for leaf meshes
   * @param leafCells - Cells per leaf mesh
   * @param topology - How nodes connect at each level
   */
  constructor(
    depth: number,
    nodesPerLevel: number,
    leafLevel: GnosticMeshLevel = 'proton',
    leafCells: number = 3,
    topology: 'ring' | 'chain' | 'complete' | 'grid' = 'ring',
  ) {
    this.depth = depth;
    this.topology = topology;
    this.root = this.buildLevel(depth, nodesPerLevel, leafLevel, leafCells, 0);
  }

  private buildLevel(
    remainingDepth: number,
    nodesPerLevel: number,
    leafLevel: GnosticMeshLevel,
    leafCells: number,
    currentDepth: number,
  ): TowerNode[] {
    const nodes: TowerNode[] = [];

    for (let i = 0; i < nodesPerLevel; i++) {
      const neighbors = this.computeNeighbors(i, nodesPerLevel);

      if (remainingDepth <= 1) {
        // Leaf: a real scheduling mesh
        nodes.push({
          id: i,
          depth: currentDepth,
          weight: 1,
          rejections: 0,
          busy: false,
          children: new SchedulingMesh(leafLevel, leafCells),
          neighbors,
        });
      } else {
        // Branch: recurse
        const children = this.buildLevel(
          remainingDepth - 1,
          nodesPerLevel,
          leafLevel,
          leafCells,
          currentDepth + 1,
        );
        nodes.push({
          id: i,
          depth: currentDepth,
          weight: 1,
          rejections: 0,
          busy: false,
          children,
          neighbors,
        });
      }
    }

    return nodes;
  }

  private computeNeighbors(i: number, n: number): number[] {
    const neighbors: number[] = [];
    if (this.topology === 'chain') {
      if (i > 0) neighbors.push(i - 1);
      if (i < n - 1) neighbors.push(i + 1);
    } else if (this.topology === 'complete') {
      for (let j = 0; j < n; j++) if (j !== i) neighbors.push(j);
    } else if (this.topology === 'grid') {
      const cols = Math.ceil(Math.sqrt(n));
      if (i % cols > 0) neighbors.push(i - 1);
      if (i % cols < cols - 1 && i + 1 < n) neighbors.push(i + 1);
      if (i - cols >= 0) neighbors.push(i - cols);
      if (i + cols < n) neighbors.push(i + cols);
    } else {
      // ring
      neighbors.push((i + 1) % n);
      neighbors.push((i - 1 + n) % n);
    }
    return neighbors;
  }

  on(listener: (depth: number, type: string, nodeId: number) => void): void {
    this.listeners.push(listener);
  }

  private emit(depth: number, type: string, nodeId: number): void {
    for (const fn of this.listeners) fn(depth, type, nodeId);
  }

  /**
   * Activate the universe.
   *
   * Drops a task at the root level. The task cascades DOWN through
   * the tower: each level forks to its children, races them, folds
   * the result, and propagates to the next node at this level.
   *
   * At the leaf level, the SchedulingMesh activates internally.
   * Results bubble back UP through the tower.
   */
  activate<T>(
    task: unknown,
    process: (vertexId: number, payload: unknown) => T | null,
  ): T[] {
    return this.activateLevel(this.root, task, process, 0);
  }

  private activateLevel<T>(
    nodes: TowerNode[],
    task: unknown,
    process: (vertexId: number, payload: unknown) => T | null,
    depth: number,
  ): T[] {
    const results: T[] = [];
    this.rounds++;

    for (const node of nodes) {
      if (node.busy) continue;
      node.state = FORK;

      this.emit(depth, 'activate', node.id);

      if (node.children instanceof SchedulingMesh) {
        // Leaf: activate the mesh
        const mesh = node.children;
        const hub = mesh.idleHubs()[0] ?? mesh.hubs[0]!;
        const trace = mesh.activate<T>(hub, task, process);
        results.push(...trace.results);
      } else {
        // Branch: recurse into children
        const childResults = this.activateLevel(node.children, task, process, depth + 1);
        results.push(...childResults);
      }

      node.state = VOID;

      // Race at this level: if we got results, propagate to neighbor
      if (results.length > 0) {
        this.emit(depth, 'race-won', node.id);

        // Reject other nodes at this level
        for (const other of nodes) {
          if (other.id !== node.id) {
            other.rejections++;
            other.weight = godFormula(this.rounds, other.rejections);
          }
        }
        break; // First result wins (race semantics)
      } else {
        node.rejections++;
        node.weight = godFormula(this.rounds, node.rejections);
        this.emit(depth, 'race-lost', node.id);
      }
    }

    return results;
  }

  /** Composite dimension of the entire tower. */
  get compositeDimension(): number {
    return this.computeDimension(this.root);
  }

  private computeDimension(nodes: TowerNode[]): number {
    if (nodes.length === 0) return 0;
    const first = nodes[0]!;
    const childDim = first.children instanceof SchedulingMesh
      ? first.children.compositeDimension
      : this.computeDimension(first.children as TowerNode[]);
    return (nodes.length - 1) * childDim + 1;
  }

  /** Total leaf meshes in the tower. */
  get totalLeaves(): number {
    return this.countLeaves(this.root);
  }

  private countLeaves(nodes: TowerNode[]): number {
    let count = 0;
    for (const node of nodes) {
      if (node.children instanceof SchedulingMesh) {
        count++;
      } else {
        count += this.countLeaves(node.children as TowerNode[]);
      }
    }
    return count;
  }

  /** Total vertices across all leaf meshes. */
  get totalVertices(): number {
    return this.countVertices(this.root);
  }

  private countVertices(nodes: TowerNode[]): number {
    let count = 0;
    for (const node of nodes) {
      if (node.children instanceof SchedulingMesh) {
        count += node.children.size;
      } else {
        count += this.countVertices(node.children as TowerNode[]);
      }
    }
    return count;
  }

  stats(): {
    depth: number;
    nodesPerLevel: number;
    totalLeaves: number;
    totalVertices: number;
    compositeDimension: number;
  } {
    return {
      depth: this.depth,
      nodesPerLevel: this.root.length,
      totalLeaves: this.totalLeaves,
      totalVertices: this.totalVertices,
      compositeDimension: this.compositeDimension,
    };
  }

  /** Reset the entire tower. */
  reset(): void {
    this.resetLevel(this.root);
    this.rounds = 0;
  }

  private resetLevel(nodes: TowerNode[]): void {
    for (const node of nodes) {
      node.weight = 1;
      node.rejections = 0;
      node.state = VOID;
      if (node.children instanceof SchedulingMesh) {
        node.children.reset();
      } else {
        this.resetLevel(node.children as TowerNode[]);
      }
    }
  }
}

/**
 * Create a universe.
 *
 * Example: createUniverse(3, 5, 'proton', 3, 'ring')
 * = 3 levels deep, 5 nodes per level, Proton leaves with 3 cells each
 * = 5^3 = 125 leaf meshes
 * = 125 × (3×1 + 3 + 1) = 125 × 7 = 875 vertices
 * = composite dim: cascading (n-1)*d + 1 three times up
 */
export function createUniverse(
  depth: number = 3,
  nodesPerLevel: number = 3,
  leafLevel: GnosticMeshLevel = 'proton',
  leafCells: number = 3,
  topology: 'ring' | 'chain' | 'complete' | 'grid' = 'ring',
): UniverseMesh {
  return new UniverseMesh(depth, nodesPerLevel, leafLevel, leafCells, topology);
}
