// @a0n/gnosis/src/seismic-stress-toy-demo.ts
/**
 * Deterministic stdout demo for `gnode run` — not imported by library consumers.
 * Kept intentionally small so `gnode run`’s static runner can analyze `main` without fallback.
 */
import {
  normalizeRateStateAsperityParams,
  normalizeSingleAsperityParams,
  simulateCoupledTwoPatch,
  simulateCoupledTwoPatchRateState,
} from './seismic-stress-toy';

/**
 * Runs the gnosis command-line workflow.
 */
export function main(): void {
  const fixed = normalizeSingleAsperityParams({
    criticalStress: 12,
    loadingRate: 1,
    stressDrop: 4,
    residualStress: 0,
    mode: 'subtract',
  });
  const coupledFixed = simulateCoupledTwoPatch(3, 8, { kappa: 0.3, patch0: fixed, patch1: fixed }, 6);
  const rs = normalizeRateStateAsperityParams({
    critical0: 20,
    b: 0.02,
    agingRate: 0.05,
    dt: 1,
    loadingRate: 0.5,
    stressDrop: 3,
    residualStress: 0,
    mode: 'subtract',
  });
  const coupledRs = simulateCoupledTwoPatchRateState(4, 9, 0.1, 0.2, { kappa: 0.25, patch0: rs, patch1: rs }, 5);
  const cf0 = coupledFixed.stresses0[6];
  const cf1 = coupledFixed.stresses1[6];
  const cr0 = coupledRs.stresses0[5];
  const cr1 = coupledRs.stresses1[5];
  console.log('seismic-stress-toy-demo');
  console.log('disclaimer: discrete toy only; not continuum seismology');
  console.log(cf0);
  console.log(cf1);
  console.log(cr0);
  console.log(cr1);
  return;
}
