// @a0n/gnosis/src/simultaneity.ts
/**
 * Simultaneity — Lorentz boosts and the relativity of simultaneity.
 *
 * A Lorentz boost along the x-axis transforms spacetime coordinates
 * between two inertial frames in relative motion. Events that are
 * simultaneous in one frame may not be simultaneous in another --
 * this module computes those breaks explicitly.
 *
 * Natural units: c = 1 lpLo/pLo (one light-picolorenzo per picolorenzo).
 * Spatial coordinates arrive in light-years and are converted to
 * light-picolorenzos for the boost, then converted back.
 *
 * @packageDocumentation
 */

import type { SpacetimeEvent } from './causal-topology';
import { LIGHT_PICOLORENZO_LY } from './math';

// ─── Types ──────────────────────────────────────────────────────────

/**
 * A spacetime event after a Lorentz boost, carrying both the
 * transformed coordinates and the boost parameters.
 */
export interface BoostedEvent extends SpacetimeEvent {
  readonly originalPLo: number;
  readonly originalX: number;
  readonly gamma: number;
  readonly beta: number;
}

/**
 * A pair of events whose simultaneity status changed under a boost.
 */
export interface SimultaneityBreak {
  readonly eventA: string;
  readonly eventB: string;
  readonly originalDtPLo: number;
  readonly boostedDtPLo: number;
}

/**
 * Full simultaneity analysis for a set of events under a Lorentz boost.
 */
export interface SimultaneitySlice {
  readonly source: 'GnosisSimultaneitySlice';
  readonly beta: number;
  readonly gamma: number;
  readonly boostedEvents: readonly BoostedEvent[];
  readonly simultaneityBreaks: readonly SimultaneityBreak[];
}

// ─── Simultaneity threshold ─────────────────────────────────────────

/** Two events count as simultaneous if |dt| < this (in picolorenzos). */
const SIMULTANEITY_EPSILON = 1e-9;

// ─── Lorentz Factor ─────────────────────────────────────────────────

/**
 * Lorentz factor gamma = 1 / sqrt(1 - beta^2).
 *
 * Throws if |beta| >= 1 (superluminal velocities are unphysical).
 */
export function lorentzGamma(beta: number): number {
  if (Math.abs(beta) >= 1) {
    throw new RangeError(
      `|beta| must be < 1 (speed of light). Got beta = ${beta}`
    );
  }
  return 1 / Math.sqrt(1 - beta * beta);
}

// ─── Single-Event Boost ─────────────────────────────────────────────

/**
 * Lorentz-boost a single spacetime event along the x-axis.
 *
 * The boost transforms from the rest frame S to a frame S' moving at
 * velocity beta (in units of c) along the x-axis:
 *
 *   t' = gamma * (t - beta * x)
 *   x' = gamma * (x - beta * t)
 *   y' = y
 *   z' = z
 *
 * x is converted from light-years to light-picolorenzos before the
 * boost and converted back afterward, so the output x is in light-years.
 * Time (pLo) stays in picolorenzos throughout.
 */
export function lorentzBoostEvent(
  event: SpacetimeEvent,
  beta: number
): BoostedEvent {
  const gamma = lorentzGamma(beta);

  // Convert x from light-years to light-picolorenzos
  const xLpLo = event.x / LIGHT_PICOLORENZO_LY;
  const t = event.pLo;

  // Apply the boost (c = 1 in natural units)
  const tPrime = gamma * (t - beta * xLpLo);
  const xPrimeLpLo = gamma * (xLpLo - beta * t);

  // Convert x' back to light-years
  const xPrimeLy = xPrimeLpLo * LIGHT_PICOLORENZO_LY;

  return {
    id: event.id,
    label: event.label,
    x: xPrimeLy,
    y: event.y,
    z: event.z,
    pLo: tPrime,
    originalPLo: event.pLo,
    originalX: event.x,
    gamma,
    beta,
  };
}

// ─── Batch Boost ────────────────────────────────────────────────────

/**
 * Lorentz-boost an array of spacetime events along the x-axis.
 */
export function lorentzBoostEvents(
  events: readonly SpacetimeEvent[],
  beta: number
): readonly BoostedEvent[] {
  return events.map((e) => lorentzBoostEvent(e, beta));
}

// ─── Simultaneity Slice ─────────────────────────────────────────────

/**
 * Compute the simultaneity slice: boost all events and identify pairs
 * whose simultaneity status changed.
 *
 * A "simultaneity break" occurs when:
 *   - Two events were simultaneous in the original frame (|dt| < epsilon)
 *     but are not simultaneous in the boosted frame, or
 *   - Two events were not simultaneous in the original frame but become
 *     simultaneous after the boost.
 */
export function simultaneitySlice(
  events: readonly SpacetimeEvent[],
  beta: number
): SimultaneitySlice {
  const gamma = lorentzGamma(beta);
  const boostedEvents = lorentzBoostEvents(events, beta);
  const breaks: SimultaneityBreak[] = [];

  for (let i = 0; i < events.length; i++) {
    for (let j = i + 1; j < events.length; j++) {
      const originalDt = Math.abs(events[i]!.pLo - events[j]!.pLo);
      const boostedDt = Math.abs(boostedEvents[i]!.pLo - boostedEvents[j]!.pLo);

      const wasSimultaneous = originalDt < SIMULTANEITY_EPSILON;
      const isSimultaneous = boostedDt < SIMULTANEITY_EPSILON;

      if (wasSimultaneous !== isSimultaneous) {
        breaks.push({
          eventA: events[i]!.id,
          eventB: events[j]!.id,
          originalDtPLo: events[i]!.pLo - events[j]!.pLo,
          boostedDtPLo: boostedEvents[i]!.pLo - boostedEvents[j]!.pLo,
        });
      }
    }
  }

  return {
    source: 'GnosisSimultaneitySlice',
    beta,
    gamma,
    boostedEvents,
    simultaneityBreaks: breaks,
  };
}
