// @a0n/gnosis/src/sofa-topology.ts
import { 
  type FlowObstruction, 
  sofaObstruction 
} from './flow-obstruction';

/**
 * The Gerver Sofa Constant (Lower Bound).
 * Representing the largest known area of a sofa that can maneuver 
 * an L-shaped corridor of unit width.
 * 
 * Approx: 2.219531669
 */
export const SOFA_CONSTANT_GERVER = 2.219531669;

/**
 * The Hammersley Sofa Constant.
 * The area of a rigid, partially convex-constrained sofa.
 * 
 * Approx: 2.2074
 */
export const SOFA_CONSTANT_HAMMERSLEY = 2.2074;

/**
 * Validates a sofa area against the Gnosis manifold benchmarks.
 */
export function validateSofaArea(area: number): { 
  isOptimalKnown: boolean;
  beatsHammersley: boolean;
  deltaToGerver: number;
} {
  return {
    isOptimalKnown: area >= SOFA_CONSTANT_GERVER,
    beatsHammersley: area > SOFA_CONSTANT_HAMMERSLEY,
    deltaToGerver: SOFA_CONSTANT_GERVER - area
  };
}

/**
 * Represents a Sofa configuration in the manifold.
 */
export interface SofaTopology {
  readonly area: number;
  readonly segments: number; // 18 for Gerver
  readonly isFlexible: boolean;
  readonly curvatureSequence?: readonly number[];
}

/**
 * Analyzes a sofa topology against a corridor width to determine 
 * if it causes a flow obstruction.
 */
export function analyzeSofaObstruction(
  topology: SofaTopology,
  corridorWidth: number
): FlowObstruction {
  return sofaObstruction(
    topology.area, 
    corridorWidth, 
    SOFA_CONSTANT_GERVER
  );
}

/**
 * The canonical Gerver Sofa (Gnosis-scaled).
 */
export const GERVER_SOFA: SofaTopology = {
  area: SOFA_CONSTANT_GERVER,
  segments: 18,
  isFlexible: true
};
