// @a0n/gnosis/src/solar-scheduler.ts
/**
 * Solar Scheduler — route work along the daylight terminator.
 *
 * Every AeonPID on Earth's surface experiences a diurnal cycle of
 * solar altitude. As the terminator sweeps west at 15°/hour, PIDs
 * wake up in the morning (load ramping, capacity expanding) and
 * wind down in the evening (load shedding, capacity contracting).
 *
 * Instead of fighting this cycle with artificial load balancing,
 * the solar scheduler ROUTES WITH IT. Work flows toward the
 * sunrise side of the network -- the PIDs that are warming up,
 * with rising capacity and declining contention. Work flows AWAY
 * from the sunset side -- the PIDs that are shedding load.
 *
 * The result: over 24 hours, work naturally circumnavigates the
 * globe, always 90 minutes ahead of local noon. The scheduler
 * becomes a sundial.
 *
 * @packageDocumentation
 */

import type { SolarObserver } from './math';
import { solarPosition } from './math';
import type { ProcessPhase } from './clock-distributed';

// ─── Types ──────────────────────────────────────────────────────────

/**
 * A node eligible to receive scheduled work.
 */
export interface SolarSchedulerNode {
  readonly pidId: string;
  readonly label: string;
  readonly lat: number;
  readonly lon: number;
  readonly capacity: number;      // max requests/sec the node can handle
  readonly currentLoad: number;   // current requests/sec
  readonly phase: ProcessPhase;   // active/idle/cold/unreachable
}

/**
 * Terminator phase direction -- where a PID sits on the diurnal cycle.
 * The key distinction is slope, not just altitude:
 *   rising: dawn through mid-morning (capacity expanding)
 *   peak:   late morning through early afternoon (capacity saturated)
 *   falling: mid-afternoon through dusk (capacity contracting)
 *   night:  sun below horizon (minimal user traffic, spare capacity)
 */
export type PhaseDirection = 'rising' | 'peak' | 'falling' | 'night';

/**
 * A scored scheduling target.
 */
export interface SolarScheduleScore {
  readonly pidId: string;
  readonly label: string;
  readonly score: number;              // 0-1, higher = better target for new work
  readonly solarAltitudeDeg: number;
  readonly solarSlopeDegPerHour: number;
  readonly phaseDirection: PhaseDirection;
  readonly capacityAvailable: number;   // capacity - currentLoad
  readonly utilizationFraction: number; // currentLoad / capacity
  readonly terminatorDistanceDeg: number; // angular distance to sunrise terminator
}

// ─── Slope Computation ──────────────────────────────────────────────

/**
 * Compute solar altitude slope (degrees per hour) at a given location and time.
 * Positive = sun rising, negative = sun setting.
 */
export function solarSlope(
  observer: SolarObserver,
  epochMs: number,
  lookbackMs: number = 600_000  // 10 minutes default
): number {
  const past = solarPosition(observer, epochMs - lookbackMs).altitudeDeg;
  const future = solarPosition(observer, epochMs + lookbackMs).altitudeDeg;
  const hoursSpan = (lookbackMs * 2) / 3_600_000;
  return (future - past) / hoursSpan;
}

/**
 * Classify a PID's phase direction from altitude and slope.
 *
 * Thresholds:
 *   altitude > 15° AND |slope| < 3°/hr → peak (midday plateau)
 *   altitude > -5° AND slope > 0 → rising (dawn through morning)
 *   altitude > -5° AND slope < 0 → falling (afternoon through dusk)
 *   altitude ≤ -5° → night
 */
export function classifyPhaseDirection(
  altitudeDeg: number,
  slopeDegPerHour: number
): PhaseDirection {
  if (altitudeDeg <= -5) return 'night';
  if (altitudeDeg > 15 && Math.abs(slopeDegPerHour) < 3) return 'peak';
  if (slopeDegPerHour > 0) return 'rising';
  if (slopeDegPerHour < 0) return 'falling';
  // Edge case: exactly 0 slope, altitude between -5 and 15
  return altitudeDeg > 0 ? 'peak' : 'night';
}

// ─── Scoring ────────────────────────────────────────────────────────

/**
 * Phase score: how attractive is this PID's diurnal position for new work?
 *
 *   rising  → 0.9 (ideal: warming up, user traffic not yet peaked)
 *   night   → 0.8 (spare capacity, local users asleep)
 *   peak    → 0.5 (saturated with local traffic, compete for resources)
 *   falling → 0.2 (shedding load, will soon go idle -- wasted warmth)
 *
 * These are *base* scores before capacity and utilization weighting.
 */
export function phaseScore(direction: PhaseDirection): number {
  switch (direction) {
    case 'rising': return 0.9;
    case 'night': return 0.8;
    case 'peak': return 0.5;
    case 'falling': return 0.2;
  }
}

/**
 * Full scheduling score combining:
 *   phase direction (0.9 rising, 0.8 night, 0.5 peak, 0.2 falling)
 *   available capacity (capacity - load) / capacity
 *   phase penalty (cold/unreachable nodes scored near zero)
 *
 * Final score = phaseScore × capacityAvailableFraction × phasePenalty
 */
export function scheduleScore(
  node: SolarSchedulerNode,
  epochMs: number,
  lookbackMs?: number
): SolarScheduleScore {
  const observer: SolarObserver = { latitudeDeg: node.lat, longitudeDeg: node.lon };
  const pos = solarPosition(observer, epochMs);
  const slope = solarSlope(observer, epochMs, lookbackMs);
  const direction = classifyPhaseDirection(pos.altitudeDeg, slope);

  const capacityAvailable = Math.max(0, node.capacity - node.currentLoad);
  const utilization = node.capacity > 0 ? node.currentLoad / node.capacity : 1;
  const capacityFraction = node.capacity > 0 ? capacityAvailable / node.capacity : 0;

  // Phase penalty for non-active processes
  const phasePenalty = node.phase === 'active' ? 1.0
    : node.phase === 'idle' ? 0.6
    : node.phase === 'cold' ? 0.1
    : 0; // unreachable

  const base = phaseScore(direction);
  const score = base * capacityFraction * phasePenalty;

  // Angular distance to sunrise terminator:
  // The sunrise terminator is where altitude = 0 with positive slope.
  // Approximate by finding where altitude would be 0 along the slope.
  const terminatorDistanceDeg = slope > 0
    ? Math.abs(pos.altitudeDeg / slope) * 15 // hours × 15°/hr
    : 360; // far from sunrise if slope non-positive

  return {
    pidId: node.pidId,
    label: node.label,
    score,
    solarAltitudeDeg: pos.altitudeDeg,
    solarSlopeDegPerHour: slope,
    phaseDirection: direction,
    capacityAvailable,
    utilizationFraction: utilization,
    terminatorDistanceDeg,
  };
}

/**
 * Score all nodes and sort by descending score (best target first).
 */
export function scoreNodes(
  nodes: readonly SolarSchedulerNode[],
  epochMs: number,
  lookbackMs?: number
): readonly SolarScheduleScore[] {
  return nodes
    .map(node => scheduleScore(node, epochMs, lookbackMs))
    .sort((a, b) => b.score - a.score);
}

// ─── Routing Decisions ──────────────────────────────────────────────

/**
 * Pick the best node to route a new unit of work.
 *
 * This is the core scheduler primitive: given a set of eligible
 * nodes and the current time, return the one most aligned with
 * the sunrise terminator.
 *
 * Returns null if no node has any available capacity.
 */
export function routeToSunrise(
  nodes: readonly SolarSchedulerNode[],
  epochMs: number = Date.now()
): SolarScheduleScore | null {
  const scored = scoreNodes(nodes, epochMs);
  const best = scored[0];
  if (!best || best.score === 0) return null;
  return best;
}

/**
 * Pick the top-K nodes along the terminator.
 * Useful for work that needs to be replicated across multiple PIDs
 * (consensus, fault tolerance, fan-out patterns).
 */
export function routeToSunriseK(
  nodes: readonly SolarSchedulerNode[],
  k: number,
  epochMs: number = Date.now()
): readonly SolarScheduleScore[] {
  return scoreNodes(nodes, epochMs)
    .filter(s => s.score > 0)
    .slice(0, Math.max(0, k));
}

// ─── Terminator Geometry ────────────────────────────────────────────

/**
 * A point on the daylight terminator line (sunrise or sunset side).
 */
export interface TerminatorPoint {
  readonly lat: number;
  readonly lon: number;
  readonly side: 'sunrise' | 'sunset';
}

/**
 * Sample the daylight terminator as a curve of lat/lon points.
 *
 * The terminator is the great circle where solar altitude = 0.
 * At each latitude, we find the longitude where the sun is exactly
 * on the horizon. Two solutions per latitude: sunrise (east side,
 * slope positive) and sunset (west side, slope negative).
 *
 * Used by the atlas visualization to draw the terminator line on the globe.
 */
export function sampleTerminator(
  epochMs: number,
  latitudeStepDeg: number = 5
): readonly TerminatorPoint[] {
  const points: TerminatorPoint[] = [];

  for (let lat = -85; lat <= 85; lat += latitudeStepDeg) {
    // Binary search for sunrise (positive slope at altitude = 0)
    const sunriseLon = findTerminatorLongitude(lat, epochMs, 'sunrise');
    if (sunriseLon !== null) {
      points.push({ lat, lon: sunriseLon, side: 'sunrise' });
    }

    const sunsetLon = findTerminatorLongitude(lat, epochMs, 'sunset');
    if (sunsetLon !== null) {
      points.push({ lat, lon: sunsetLon, side: 'sunset' });
    }
  }

  return points;
}

/**
 * Find the longitude where the solar altitude crosses zero at a given
 * latitude, on the requested side (sunrise = positive slope, sunset = negative).
 *
 * Uses ternary search: step through longitudes and find where altitude
 * sign changes in the correct direction.
 */
function findTerminatorLongitude(
  lat: number,
  epochMs: number,
  side: 'sunrise' | 'sunset',
  step: number = 2
): number | null {
  let prevAlt = solarPosition({ latitudeDeg: lat, longitudeDeg: -180 }, epochMs).altitudeDeg;

  for (let lon = -180 + step; lon <= 180; lon += step) {
    const alt = solarPosition({ latitudeDeg: lat, longitudeDeg: lon }, epochMs).altitudeDeg;

    // Zero crossing detection
    if (Math.sign(prevAlt) !== Math.sign(alt) && prevAlt !== 0) {
      const rising = alt > prevAlt;
      const match = (side === 'sunrise' && rising) || (side === 'sunset' && !rising);

      if (match) {
        // Linear interpolation for precise crossing
        const t = Math.abs(prevAlt) / (Math.abs(prevAlt) + Math.abs(alt));
        return lon - step + step * t;
      }
    }

    prevAlt = alt;
  }

  return null;
}

// ─── Work Flow Simulation ───────────────────────────────────────────

/**
 * Simulate how scheduled work distributes across nodes over a time window.
 * Useful for capacity planning and visualization.
 */
export interface WorkFlowStep {
  readonly epochMs: number;
  readonly targetPidId: string;
  readonly targetLabel: string;
  readonly score: number;
  readonly phaseDirection: PhaseDirection;
}

/**
 * Handles the gnosis simulate Work Flow workflow.
 */
export function simulateWorkFlow(
  nodes: readonly SolarSchedulerNode[],
  startEpochMs: number,
  endEpochMs: number,
  stepMs: number = 3_600_000 // hourly
): readonly WorkFlowStep[] {
  const steps: WorkFlowStep[] = [];

  for (let t = startEpochMs; t <= endEpochMs; t += stepMs) {
    const best = routeToSunrise(nodes, t);
    if (best) {
      steps.push({
        epochMs: t,
        targetPidId: best.pidId,
        targetLabel: best.label,
        score: best.score,
        phaseDirection: best.phaseDirection,
      });
    }
  }

  return steps;
}

/**
 * Distribution summary: how did work spread across PIDs during a simulation?
 * Shows which nodes received the most work.
 */
export interface WorkFlowDistribution {
  readonly source: 'GnosisSolarSchedulerDistribution';
  readonly totalSteps: number;
  readonly perNodeCount: ReadonlyMap<string, number>;
  readonly perNodeFraction: ReadonlyMap<string, number>;
  readonly dominantPidId: string | null;
}

/**
 * Handles the gnosis summarize Work Flow workflow.
 */
export function summarizeWorkFlow(
  steps: readonly WorkFlowStep[]
): WorkFlowDistribution {
  const counts = new Map<string, number>();
  for (const step of steps) {
    counts.set(step.targetPidId, (counts.get(step.targetPidId) ?? 0) + 1);
  }

  const total = steps.length;
  const fractions = new Map<string, number>();
  for (const [id, count] of counts) {
    fractions.set(id, total > 0 ? count / total : 0);
  }

  let dominant: string | null = null;
  let maxCount = 0;
  for (const [id, count] of counts) {
    if (count > maxCount) {
      maxCount = count;
      dominant = id;
    }
  }

  return {
    source: 'GnosisSolarSchedulerDistribution',
    totalSteps: total,
    perNodeCount: counts,
    perNodeFraction: fractions,
    dominantPidId: dominant,
  };
}
