// @a0n/gnosis/src/steering-trace.ts
import * as os from 'os';
import * as path from 'path';
import { createHash } from 'node:crypto';
import twokeys from '@a0n/twokeys';
import type { PointsDescription, SeriesDescription } from '@a0n/twokeys';
import {
  QDoc,
  QDocRelay,
  type QDocEvent,
  type QDocRelayConfig,
  type QDocRelayStatus,
} from './crdt/index';
import {
  authorizeSteeringRelayConnect,
  authorizeSteeringRun,
  authorizeSteeringTraceAppend,
  normalizeExecutionAuthContext,
  type GnosisExecutionAuthContext,
  type GnosisSteeringAuthorizationResult,
} from './auth/core';
import { classifySteeringBias, type GnosisSteeringBias } from './analysis';
import type {
  GnosisCollapseCostAction,
  GnosisComplexityReport,
  GnosisSteeringAction,
  GnosisSteeringMetrics,
  GnosisSteeringMode,
  GnosisSteeringRegime,
  GnosisSteeringTelemetry,
} from './analysis';

const { Points, Series } = twokeys;

export const DEFAULT_GNOSIS_STEERING_TRACE_WINDOW = 256;

export type GnosisSteeringTraceCommand = 'run' | 'native' | 'repl';
export type GnosisSteeringTraceOutcome = 'ok' | 'error';
export const DEFAULT_GNOSIS_STEERING_TRACE_GUID = 'gnosis-steering-trace';

export interface GnosisSteeringTraceEnvironment {
  runtime: string;
  runtimeVersion: string;
  platform: NodeJS.Platform;
  arch: string;
  hostFingerprint: string;
}

export interface GnosisSteeringTraceRecord {
  version: 2;
  traceId: string;
  recordedAt: string;
  command: GnosisSteeringTraceCommand;
  steeringMode: GnosisSteeringMode;
  topologyPath: string | null;
  topologyName: string | null;
  topologyFingerprint: string;
  cohortKey: string;
  workloadKey: string;
  environment: GnosisSteeringTraceEnvironment;
  buleyNumber: number;
  wallaceNumber: number;
  wallaceUnit: 'wally';
  topologyDeficit: number;
  frontierFill: number;
  frontierDeficit: number;
  regime: GnosisSteeringRegime;
  boundaryWalkBias: number;
  leanInBias: GnosisSteeringBias;
  recommendedAction: GnosisSteeringAction | null;
  collapseCostAction: GnosisCollapseCostAction;
  totalVentCost: number;
  totalRepairDebt: number;
  totalPaidCost: number;
  paidStageCount: number;
  collapseCostFloor: number;
  collapseCostMargin: number;
  prefixCostDeficit: number;
  zeroWasteCollapseRisk: boolean;
  contagiousRepairPressure: boolean;
  freeCollapsePrefixRisk: boolean;
  telemetry: GnosisSteeringTelemetry;
  stateCount: number;
  exploredBeta1: number;
  structuralBeta1: number;
  graphDensity: number | null;
  graphDiameter: number | null;
  payloadBytes: number;
  payloadFingerprint: string;
  outcome: GnosisSteeringTraceOutcome;
  errorMessage: string | null;
}

export interface GnosisSteeringTraceSummary {
  recordCount: number;
  topologyCount: number;
  hostCount: number;
  latestRecordedAt: string | null;
  latestFailureRecordedAt: string | null;
  latestFailureMessage: string | null;
  cohortKey: string | null;
  invalidLineCount: number;
  truncated: boolean;
  recentWindow: number | null;
  successCount: number;
  failureCount: number;
  failureRate: number;
  alertLevel: 'healthy' | 'warning' | 'critical';
  wallace: SeriesDescription | null;
  boundaryWalkBias: SeriesDescription | null;
  leanInCounts: Record<GnosisSteeringBias, number>;
  wallMicroCharleys: SeriesDescription | null;
  cpuMicroCharleys: SeriesDescription | null;
  wallToCpuRatio: SeriesDescription | null;
  wallaceVsWallMicroCharleys: PointsDescription | null;
  wallaceVsCpuMicroCharleys: PointsDescription | null;
  wallaceWallCpu: PointsDescription | null;
  wallaceToWallCorrelation: number | null;
  wallaceToCpuCorrelation: number | null;
}

export interface CreateGnosisSteeringTraceRecordOptions {
  command: GnosisSteeringTraceCommand;
  filePath: string | null;
  source: string;
  report: GnosisComplexityReport;
  steering: GnosisSteeringMetrics;
  payload: unknown;
  outcome?: GnosisSteeringTraceOutcome;
  errorMessage?: string | null;
}

export interface GnosisSteeringTraceReadOptions {
  cohortKey?: string | null;
  maxRecords?: number;
}

export interface GnosisSteeringTraceDataset {
  records: GnosisSteeringTraceRecord[];
  invalidLineCount: number;
  truncated: boolean;
  recentWindow: number | null;
  cohortKey: string | null;
}

export interface GnosisSteeringTraceStoreConfig {
  guid?: string;
  recentWindow?: number;
  relay?: QDocRelayConfig | null;
  executionAuth?: GnosisExecutionAuthContext | null;
}

export interface GnosisSteeringTraceStoreStatus {
  guid: string;
  durable: boolean;
  relayConfigured: boolean;
  relay: QDocRelayStatus | null;
  localRecordCount: number;
  cohortCount: number;
  invalidEventCount: number;
  lastRelayError: string | null;
  authEnforced: boolean;
  principal: string | null;
}

function round2(value: number): number {
  return Math.round(value * 100) / 100;
}

function stableHash(value: string): string {
  return createHash('sha256').update(value).digest('hex').slice(0, 16);
}

/**
 * Handles the gnosis fingerprint Steering Topology workflow.
 */
export function fingerprintSteeringTopology(source: string): string {
  return stableHash(source);
}

/**
 * Handles the gnosis default Steering Trace Room Name workflow.
 */
export function defaultSteeringTraceRoomName(
  workspace: string = process.cwd()
): string {
  return `gnosis:steering:${stableHash(workspace)}`;
}

function toFiniteNumber(value: number | null | undefined): number | null {
  return typeof value === 'number' && Number.isFinite(value) ? value : null;
}

function describeSeries(values: number[]): SeriesDescription | null {
  return values.length === 0 ? null : new Series({ data: values }).describe();
}

function describePoints(values: number[][]): PointsDescription | null {
  return values.length === 0 ? null : new Points({ data: values }).describe();
}

function correlationAt(
  description: PointsDescription | null,
  rowIndex: number,
  columnIndex: number
): number | null {
  const rawValue = description?.correlationMatrix[rowIndex]?.[columnIndex];
  return toFiniteNumber(rawValue);
}

function serializePayload(payload: unknown): string {
  if (typeof payload === 'string') {
    return payload;
  }

  try {
    return JSON.stringify(payload) ?? String(payload);
  } catch {
    return String(payload);
  }
}

function deriveEnvironment(): GnosisSteeringTraceEnvironment {
  return {
    runtime: process.release.name,
    runtimeVersion: process.version,
    platform: process.platform,
    arch: process.arch,
    hostFingerprint: stableHash(
      `${os.hostname()}:${process.platform}:${process.arch}:${
        process.release.name
      }`
    ),
  };
}

function deriveWorkloadKey(
  payloadFingerprint: string,
  payloadBytes: number
): string {
  return `${payloadFingerprint}:${payloadBytes}`;
}

function deriveCohortKey(
  topologyFingerprint: string,
  environment: GnosisSteeringTraceEnvironment,
  workloadKey: string,
  command: GnosisSteeringTraceCommand
): string {
  return [
    topologyFingerprint,
    workloadKey,
    command,
    environment.runtime,
    environment.platform,
    environment.arch,
  ].join(':');
}

/**
 * Handles the gnosis authorize Boundary Walk Run workflow.
 */
export function authorizeBoundaryWalkRun(options: {
  source: string;
  executionAuth: GnosisExecutionAuthContext;
}): GnosisSteeringAuthorizationResult {
  return authorizeSteeringRun({
    topologyFingerprint: fingerprintSteeringTopology(options.source),
    auth: options.executionAuth,
  });
}

function normalizeTraceRecord(
  candidate: unknown
): GnosisSteeringTraceRecord | null {
  if (!candidate || typeof candidate !== 'object') {
    return null;
  }

  const rawRecord = candidate as Record<string, unknown>;
  const telemetryCandidate =
    rawRecord.telemetry && typeof rawRecord.telemetry === 'object'
      ? (rawRecord.telemetry as Record<string, unknown>)
      : {};
  const environmentCandidate =
    rawRecord.environment && typeof rawRecord.environment === 'object'
      ? (rawRecord.environment as Record<string, unknown>)
      : {};

  const topologyFingerprint =
    typeof rawRecord.topologyFingerprint === 'string'
      ? rawRecord.topologyFingerprint
      : null;
  const recordedAt =
    typeof rawRecord.recordedAt === 'string' ? rawRecord.recordedAt : null;
  const command =
    rawRecord.command === 'run' ||
    rawRecord.command === 'native' ||
    rawRecord.command === 'repl'
      ? rawRecord.command
      : null;
  const steeringMode =
    rawRecord.steeringMode === 'off' ||
    rawRecord.steeringMode === 'report' ||
    rawRecord.steeringMode === 'suggest' ||
    rawRecord.steeringMode === 'apply'
      ? rawRecord.steeringMode
      : 'suggest';

  if (!topologyFingerprint || !recordedAt || !command) {
    return null;
  }

  const payloadFingerprint =
    typeof rawRecord.payloadFingerprint === 'string'
      ? rawRecord.payloadFingerprint
      : 'legacy';
  const payloadBytes =
    typeof rawRecord.payloadBytes === 'number' &&
    Number.isFinite(rawRecord.payloadBytes)
      ? rawRecord.payloadBytes
      : 0;
  const environment: GnosisSteeringTraceEnvironment = {
    runtime:
      typeof environmentCandidate.runtime === 'string'
        ? environmentCandidate.runtime
        : 'legacy',
    runtimeVersion:
      typeof environmentCandidate.runtimeVersion === 'string'
        ? environmentCandidate.runtimeVersion
        : 'legacy',
    platform:
      typeof environmentCandidate.platform === 'string'
        ? (environmentCandidate.platform as NodeJS.Platform)
        : 'linux',
    arch:
      typeof environmentCandidate.arch === 'string'
        ? environmentCandidate.arch
        : 'unknown',
    hostFingerprint:
      typeof environmentCandidate.hostFingerprint === 'string'
        ? environmentCandidate.hostFingerprint
        : 'legacy',
  };
  const workloadKey =
    typeof rawRecord.workloadKey === 'string'
      ? rawRecord.workloadKey
      : deriveWorkloadKey(payloadFingerprint, payloadBytes);
  const cohortKey =
    typeof rawRecord.cohortKey === 'string'
      ? rawRecord.cohortKey
      : deriveCohortKey(topologyFingerprint, environment, workloadKey, command);

  return {
    version: 2,
    traceId:
      typeof rawRecord.traceId === 'string'
        ? rawRecord.traceId
        : crypto.randomUUID(),
    recordedAt,
    command,
    steeringMode,
    topologyPath:
      typeof rawRecord.topologyPath === 'string'
        ? rawRecord.topologyPath
        : null,
    topologyName:
      typeof rawRecord.topologyName === 'string'
        ? rawRecord.topologyName
        : null,
    topologyFingerprint,
    cohortKey,
    workloadKey,
    environment,
    buleyNumber:
      typeof rawRecord.buleyNumber === 'number' &&
      Number.isFinite(rawRecord.buleyNumber)
        ? rawRecord.buleyNumber
        : 0,
    wallaceNumber:
      typeof rawRecord.wallaceNumber === 'number' &&
      Number.isFinite(rawRecord.wallaceNumber)
        ? rawRecord.wallaceNumber
        : 0,
    wallaceUnit: 'wally',
    topologyDeficit:
      typeof rawRecord.topologyDeficit === 'number' &&
      Number.isFinite(rawRecord.topologyDeficit)
        ? rawRecord.topologyDeficit
        : 0,
    frontierFill:
      typeof rawRecord.frontierFill === 'number' &&
      Number.isFinite(rawRecord.frontierFill)
        ? rawRecord.frontierFill
        : 0,
    frontierDeficit:
      typeof rawRecord.frontierDeficit === 'number' &&
      Number.isFinite(rawRecord.frontierDeficit)
        ? rawRecord.frontierDeficit
        : 0,
    regime:
      rawRecord.regime === 'laminar' ||
      rawRecord.regime === 'transitional' ||
      rawRecord.regime === 'turbulent'
        ? rawRecord.regime
        : 'laminar',
    boundaryWalkBias:
      typeof rawRecord.boundaryWalkBias === 'number' &&
      Number.isFinite(rawRecord.boundaryWalkBias)
        ? rawRecord.boundaryWalkBias
        : 0,
    leanInBias:
      rawRecord.leanInBias === 'lean-in' ||
      rawRecord.leanInBias === 'neutral' ||
      rawRecord.leanInBias === 'pull-back'
        ? rawRecord.leanInBias
        : classifySteeringBias(
            typeof rawRecord.boundaryWalkBias === 'number' &&
              Number.isFinite(rawRecord.boundaryWalkBias)
              ? rawRecord.boundaryWalkBias
              : 0
          ),
    recommendedAction:
      rawRecord.recommendedAction === 'expand' ||
      rawRecord.recommendedAction === 'staggered-expand' ||
      rawRecord.recommendedAction === 'hold' ||
      rawRecord.recommendedAction === 'multiplex' ||
      rawRecord.recommendedAction === 'constrain'
        ? rawRecord.recommendedAction
        : null,
    collapseCostAction:
      rawRecord.collapseCostAction === 'none' ||
      rawRecord.collapseCostAction === 'keep-multiplicity' ||
      rawRecord.collapseCostAction === 'pay-vent' ||
      rawRecord.collapseCostAction === 'pay-repair' ||
      rawRecord.collapseCostAction === 'pay-vent-or-repair'
        ? rawRecord.collapseCostAction
        : 'none',
    totalVentCost:
      typeof rawRecord.totalVentCost === 'number' &&
      Number.isFinite(rawRecord.totalVentCost)
        ? rawRecord.totalVentCost
        : 0,
    totalRepairDebt:
      typeof rawRecord.totalRepairDebt === 'number' &&
      Number.isFinite(rawRecord.totalRepairDebt)
        ? rawRecord.totalRepairDebt
        : 0,
    totalPaidCost:
      typeof rawRecord.totalPaidCost === 'number' &&
      Number.isFinite(rawRecord.totalPaidCost)
        ? rawRecord.totalPaidCost
        : 0,
    paidStageCount:
      typeof rawRecord.paidStageCount === 'number' &&
      Number.isFinite(rawRecord.paidStageCount)
        ? rawRecord.paidStageCount
        : 0,
    collapseCostFloor:
      typeof rawRecord.collapseCostFloor === 'number' &&
      Number.isFinite(rawRecord.collapseCostFloor)
        ? rawRecord.collapseCostFloor
        : 0,
    collapseCostMargin:
      typeof rawRecord.collapseCostMargin === 'number' &&
      Number.isFinite(rawRecord.collapseCostMargin)
        ? rawRecord.collapseCostMargin
        : 0,
    prefixCostDeficit:
      typeof rawRecord.prefixCostDeficit === 'number' &&
      Number.isFinite(rawRecord.prefixCostDeficit)
        ? rawRecord.prefixCostDeficit
        : 0,
    zeroWasteCollapseRisk: rawRecord.zeroWasteCollapseRisk === true,
    contagiousRepairPressure: rawRecord.contagiousRepairPressure === true,
    freeCollapsePrefixRisk: rawRecord.freeCollapsePrefixRisk === true,
    telemetry: {
      wallMicroCharleys: toFiniteNumber(
        telemetryCandidate.wallMicroCharleys as number | null | undefined
      ),
      cpuMicroCharleys: toFiniteNumber(
        telemetryCandidate.cpuMicroCharleys as number | null | undefined
      ),
      wallToCpuRatio: toFiniteNumber(
        telemetryCandidate.wallToCpuRatio as number | null | undefined
      ),
    },
    stateCount:
      typeof rawRecord.stateCount === 'number' &&
      Number.isFinite(rawRecord.stateCount)
        ? rawRecord.stateCount
        : 0,
    exploredBeta1:
      typeof rawRecord.exploredBeta1 === 'number' &&
      Number.isFinite(rawRecord.exploredBeta1)
        ? rawRecord.exploredBeta1
        : 0,
    structuralBeta1:
      typeof rawRecord.structuralBeta1 === 'number' &&
      Number.isFinite(rawRecord.structuralBeta1)
        ? rawRecord.structuralBeta1
        : 0,
    graphDensity: toFiniteNumber(
      rawRecord.graphDensity as number | null | undefined
    ),
    graphDiameter: toFiniteNumber(
      rawRecord.graphDiameter as number | null | undefined
    ),
    payloadBytes,
    payloadFingerprint,
    outcome: rawRecord.outcome === 'error' ? 'error' : 'ok',
    errorMessage:
      typeof rawRecord.errorMessage === 'string'
        ? rawRecord.errorMessage
        : null,
  };
}

/**
 * Creates the Steering Trace Record.
 */
export function createSteeringTraceRecord(
  options: CreateGnosisSteeringTraceRecordOptions
): GnosisSteeringTraceRecord {
  const serializedPayload = serializePayload(options.payload);
  const payloadBytes = Buffer.byteLength(serializedPayload, 'utf8');
  const payloadFingerprint = stableHash(serializedPayload);
  const environment = deriveEnvironment();
  const workloadKey = deriveWorkloadKey(payloadFingerprint, payloadBytes);
  const topologyFingerprint = fingerprintSteeringTopology(options.source);

  return {
    version: 2,
    traceId: crypto.randomUUID(),
    recordedAt: new Date().toISOString(),
    command: options.command,
    steeringMode: options.steering.mode,
    topologyPath: options.filePath,
    topologyName: options.filePath ? path.basename(options.filePath) : null,
    topologyFingerprint,
    cohortKey: deriveCohortKey(
      topologyFingerprint,
      environment,
      workloadKey,
      options.command
    ),
    workloadKey,
    environment,
    buleyNumber: options.report.buleyNumber,
    wallaceNumber: options.steering.wallaceNumber,
    wallaceUnit: options.steering.wallaceUnit,
    topologyDeficit: options.steering.topologyDeficit,
    frontierFill: options.steering.frontierFill,
    frontierDeficit: options.steering.frontierDeficit,
    regime: options.steering.regime,
    boundaryWalkBias: options.steering.boundaryWalkBias,
    leanInBias: options.steering.leanInBias,
    recommendedAction: options.steering.recommendedAction,
    collapseCostAction: options.steering.failureBoundary.collapseCostAction,
    totalVentCost: options.steering.failureBoundary.totalVentCost,
    totalRepairDebt: options.steering.failureBoundary.totalRepairDebt,
    totalPaidCost: options.steering.failureBoundary.totalPaidCost,
    paidStageCount: options.steering.failureBoundary.paidStageCount,
    collapseCostFloor: options.steering.failureBoundary.collapseCostFloor,
    collapseCostMargin: options.steering.failureBoundary.collapseCostMargin,
    prefixCostDeficit: options.steering.failureBoundary.prefixCostDeficit,
    zeroWasteCollapseRisk:
      options.steering.failureBoundary.zeroWasteCollapseRisk,
    contagiousRepairPressure:
      options.steering.failureBoundary.contagiousRepairPressure,
    freeCollapsePrefixRisk:
      options.steering.failureBoundary.freeCollapsePrefixRisk,
    telemetry: options.steering.telemetry,
    stateCount: options.report.correctness.stateCount,
    exploredBeta1: options.report.correctness.topology.beta1,
    structuralBeta1: options.report.topology.structuralBeta1,
    graphDensity: toFiniteNumber(
      options.steering.eda.graph.density as number | null | undefined
    ),
    graphDiameter: toFiniteNumber(
      options.steering.eda.graph.diameter as number | null | undefined
    ),
    payloadBytes,
    payloadFingerprint,
    outcome: options.outcome ?? 'ok',
    errorMessage: options.errorMessage ?? null,
  };
}

/**
 * Handles the gnosis summarize Steering Trace Records workflow.
 */
export function summarizeSteeringTraceRecords(
  records: readonly GnosisSteeringTraceRecord[],
  options: Partial<
    Pick<
      GnosisSteeringTraceSummary,
      'invalidLineCount' | 'truncated' | 'recentWindow' | 'cohortKey'
    >
  > = {}
): GnosisSteeringTraceSummary {
  const wallaceValues = records.map((record) => record.wallaceNumber);
  const boundaryBiasValues = records.map((record) => record.boundaryWalkBias);
  const wallValues = records
    .map((record) => toFiniteNumber(record.telemetry.wallMicroCharleys))
    .filter((value): value is number => value !== null);
  const cpuValues = records
    .map((record) => toFiniteNumber(record.telemetry.cpuMicroCharleys))
    .filter((value): value is number => value !== null);
  const ratioValues = records
    .map((record) => toFiniteNumber(record.telemetry.wallToCpuRatio))
    .filter((value): value is number => value !== null);
  const wallaceVsWall = records
    .map((record) => {
      const wallMicroCharleys = toFiniteNumber(
        record.telemetry.wallMicroCharleys
      );
      return wallMicroCharleys === null
        ? null
        : [record.wallaceNumber, wallMicroCharleys];
    })
    .filter((value): value is [number, number] => value !== null);
  const wallaceVsCpu = records
    .map((record) => {
      const cpuMicroCharleys = toFiniteNumber(
        record.telemetry.cpuMicroCharleys
      );
      return cpuMicroCharleys === null
        ? null
        : [record.wallaceNumber, cpuMicroCharleys];
    })
    .filter((value): value is [number, number] => value !== null);
  const wallaceWallCpu = records
    .map((record) => {
      const wallMicroCharleys = toFiniteNumber(
        record.telemetry.wallMicroCharleys
      );
      const cpuMicroCharleys = toFiniteNumber(
        record.telemetry.cpuMicroCharleys
      );

      return wallMicroCharleys === null || cpuMicroCharleys === null
        ? null
        : [record.wallaceNumber, wallMicroCharleys, cpuMicroCharleys];
    })
    .filter((value): value is [number, number, number] => value !== null);
  const wallaceVsWallMicroCharleys = describePoints(wallaceVsWall);
  const wallaceVsCpuMicroCharleys = describePoints(wallaceVsCpu);
  const successCount = records.filter(
    (record) => record.outcome === 'ok'
  ).length;
  const failureCount = records.length - successCount;
  const leanInCounts: Record<GnosisSteeringBias, number> = {
    'lean-in': records.filter((record) => record.leanInBias === 'lean-in')
      .length,
    neutral: records.filter((record) => record.leanInBias === 'neutral').length,
    'pull-back': records.filter((record) => record.leanInBias === 'pull-back')
      .length,
  };
  const latestFailure =
    [...records].reverse().find((record) => record.outcome === 'error') ?? null;
  const invalidLineCount = options.invalidLineCount ?? 0;
  const failureRate =
    records.length === 0 ? 0 : round2((failureCount / records.length) * 100);
  const alertLevel =
    failureCount > 0
      ? 'critical'
      : invalidLineCount > 0
      ? 'warning'
      : 'healthy';

  return {
    recordCount: records.length,
    topologyCount: new Set(records.map((record) => record.topologyFingerprint))
      .size,
    hostCount: new Set(
      records.map((record) => record.environment.hostFingerprint)
    ).size,
    latestRecordedAt:
      records.length === 0
        ? null
        : records[records.length - 1]?.recordedAt ?? null,
    latestFailureRecordedAt: latestFailure?.recordedAt ?? null,
    latestFailureMessage: latestFailure?.errorMessage ?? null,
    cohortKey: options.cohortKey ?? null,
    invalidLineCount,
    truncated: options.truncated ?? false,
    recentWindow: options.recentWindow ?? null,
    successCount,
    failureCount,
    failureRate,
    alertLevel,
    wallace: describeSeries(wallaceValues),
    boundaryWalkBias: describeSeries(boundaryBiasValues),
    leanInCounts,
    wallMicroCharleys: describeSeries(wallValues),
    cpuMicroCharleys: describeSeries(cpuValues),
    wallToCpuRatio: describeSeries(ratioValues),
    wallaceVsWallMicroCharleys,
    wallaceVsCpuMicroCharleys,
    wallaceWallCpu: describePoints(wallaceWallCpu),
    wallaceToWallCorrelation: correlationAt(wallaceVsWallMicroCharleys, 0, 1),
    wallaceToCpuCorrelation: correlationAt(wallaceVsCpuMicroCharleys, 0, 1),
  };
}

/**
 * Formats the Steering Trace Summary.
 */
export function formatSteeringTraceSummary(
  summary: GnosisSteeringTraceSummary
): string {
  if (summary.recordCount === 0) {
    return 'runs=0';
  }

  const parts = [
    `runs=${summary.recordCount}`,
    `topologies=${summary.topologyCount}`,
    `hosts=${summary.hostCount}`,
  ];

  if (summary.alertLevel !== 'healthy') {
    parts.push(`alert=${summary.alertLevel}`);
  }

  if (summary.failureCount > 0) {
    parts.push(`failures=${summary.failureCount}`);
    parts.push(`failure-rate=${summary.failureRate}%`);
  }

  if (summary.invalidLineCount > 0) {
    parts.push(`invalid=${summary.invalidLineCount}`);
  }

  if (summary.truncated && summary.recentWindow !== null) {
    parts.push(`window=recent-${summary.recentWindow}`);
  }

  if (summary.wallace) {
    parts.push(`wallace-median=${summary.wallace.summary.median.datum}`);
  }

  if (summary.boundaryWalkBias) {
    parts.push(`bias-median=${summary.boundaryWalkBias.summary.median.datum}`);
  }

  if (summary.leanInCounts['lean-in'] > 0) {
    parts.push(`lean-in=${summary.leanInCounts['lean-in']}`);
  }

  if (summary.leanInCounts['pull-back'] > 0) {
    parts.push(`pull-back=${summary.leanInCounts['pull-back']}`);
  }

  if (summary.wallMicroCharleys) {
    parts.push(
      `wall-uCharleys-median=${summary.wallMicroCharleys.summary.median.datum}`
    );
  }

  if (summary.cpuMicroCharleys) {
    parts.push(
      `cpu-uCharleys-median=${summary.cpuMicroCharleys.summary.median.datum}`
    );
  }

  if (summary.wallToCpuRatio) {
    parts.push(
      `wall/cpu-median=${summary.wallToCpuRatio.summary.median.datum}`
    );
  }

  const wallaceToWallCorrelation = toFiniteNumber(
    summary.wallaceToWallCorrelation
  );
  if (wallaceToWallCorrelation !== null) {
    parts.push(`wallace~wall=${round2(wallaceToWallCorrelation)}`);
  }

  const wallaceToCpuCorrelation = toFiniteNumber(
    summary.wallaceToCpuCorrelation
  );
  if (wallaceToCpuCorrelation !== null) {
    parts.push(`wallace~cpu=${round2(wallaceToCpuCorrelation)}`);
  }

  if (summary.cohortKey) {
    parts.push(`cohort=${summary.cohortKey.slice(0, 12)}`);
  }

  return parts.join(' ');
}

/**
 * Formats the Steering Trace Alert.
 */
export function formatSteeringTraceAlert(
  summary: GnosisSteeringTraceSummary,
  status?: GnosisSteeringTraceStoreStatus | null
): string | null {
  const relayDisconnected = status?.relayConfigured && !status.relay?.connected;
  const relayError = status?.lastRelayError ?? null;
  const critical =
    summary.failureCount > 0 || relayDisconnected || relayError !== null;
  const warning = summary.invalidLineCount > 0;
  const alertLevel = critical ? 'critical' : warning ? 'warning' : 'healthy';

  if (alertLevel === 'healthy') {
    return null;
  }

  const parts = [`STEERING ALERT level=${alertLevel.toUpperCase()}`];

  if (status) {
    parts.push(`durable=${status.durable}`);
    if (status.relayConfigured) {
      parts.push(
        `relay=${status.relay?.connected ? 'connected' : 'disconnected'}`
      );
      if (status.relay?.roomName) {
        parts.push(`room=${status.relay.roomName}`);
      }
      if (status.relay?.relayProduct) {
        parts.push(`product=${status.relay.relayProduct}`);
      }
      if (status.relay?.protocol) {
        parts.push(`protocol=${status.relay.protocol}`);
      }
    }
  }

  if (summary.failureCount > 0) {
    parts.push(`failures=${summary.failureCount}`);
  }

  if (summary.invalidLineCount > 0) {
    parts.push(`invalid-events=${summary.invalidLineCount}`);
  }

  if (summary.cohortKey) {
    parts.push(`cohort=${summary.cohortKey.slice(0, 12)}`);
  }

  if (summary.latestFailureMessage) {
    parts.push(`latest="${summary.latestFailureMessage}"`);
  }

  if (relayError) {
    parts.push(`relay-error="${relayError}"`);
  }

  return parts.join(' ');
}

/**
 * Formats the Steering Trace Store Status.
 */
export function formatSteeringTraceStoreStatus(
  status: GnosisSteeringTraceStoreStatus
): string {
  const parts = [
    `durable=${status.durable}`,
    `relay-configured=${status.relayConfigured}`,
    `auth-enforced=${status.authEnforced}`,
    `records=${status.localRecordCount}`,
    `cohorts=${status.cohortCount}`,
  ];

  if (status.principal) {
    parts.push(`principal=${status.principal}`);
  }

  if (status.relay) {
    parts.push(`relay-connected=${status.relay.connected}`);
    parts.push(`room=${status.relay.roomName}`);
    parts.push(`product=${status.relay.relayProduct}`);
    parts.push(`protocol=${status.relay.protocol}`);
  }

  if (status.invalidEventCount > 0) {
    parts.push(`invalid-events=${status.invalidEventCount}`);
  }

  return parts.join(' ');
}

function compareTraceRecordOrder(
  left: GnosisSteeringTraceRecord,
  right: GnosisSteeringTraceRecord
): number {
  const byRecordedAt = left.recordedAt.localeCompare(right.recordedAt);
  return byRecordedAt !== 0
    ? byRecordedAt
    : left.traceId.localeCompare(right.traceId);
}

export class GnosisSteeringTraceStore {
  private readonly doc: QDoc;
  private readonly relay: QDocRelay | null;
  private readonly recentWindow: number;
  private readonly recordsMap;
  private readonly recordsById = new Map<string, GnosisSteeringTraceRecord>();
  private readonly recordOrder: string[] = [];
  private readonly cohortRecordOrder = new Map<string, string[]>();
  private invalidEventCount = 0;
  private lastRelayError: string | null = null;
  private executionAuth: GnosisExecutionAuthContext | null = null;
  private readonly tracePath = 'steering.trace.records';
  private readonly observeHandler = (event: QDocEvent): void => {
    if (
      event.path !== this.tracePath ||
      !event.key ||
      event.value === undefined
    ) {
      return;
    }

    if (event.type === 'delete') {
      this.deleteRecord(event.key);
      return;
    }

    if (typeof event.value !== 'string') {
      this.invalidEventCount += 1;
      return;
    }

    try {
      const parsed = JSON.parse(event.value) as unknown;
      const normalizedRecord = normalizeTraceRecord(parsed);
      if (!normalizedRecord) {
        this.invalidEventCount += 1;
        return;
      }

      this.upsertRecord(normalizedRecord);
    } catch {
      this.invalidEventCount += 1;
    }
  };

  constructor(config: GnosisSteeringTraceStoreConfig = {}) {
    this.doc = new QDoc({
      guid: config.guid ?? DEFAULT_GNOSIS_STEERING_TRACE_GUID,
    });
    this.recentWindow =
      config.recentWindow ?? DEFAULT_GNOSIS_STEERING_TRACE_WINDOW;
    this.recordsMap = this.doc.getMap<string>(this.tracePath);
    this.doc.observe(this.tracePath, this.observeHandler);
    this.relay = config.relay ? new QDocRelay(this.doc, config.relay) : null;
    this.executionAuth = normalizeExecutionAuthContext(config.executionAuth);
  }

  async connect(): Promise<void> {
    if (!this.relay) {
      return;
    }

    try {
      this.assertRelayConnectAuthorized();
      await this.relay.connect();
      this.lastRelayError = null;
    } catch (error) {
      this.lastRelayError =
        error instanceof Error ? error.message : String(error);
      throw error;
    }
  }

  disconnect(): void {
    this.relay?.disconnect();
    this.doc.unobserve(this.tracePath, this.observeHandler);
    this.doc.destroy();
  }

  setExecutionAuth(executionAuth: GnosisExecutionAuthContext | null): void {
    this.executionAuth = normalizeExecutionAuthContext(executionAuth);
  }

  append(record: GnosisSteeringTraceRecord): void {
    this.assertTraceAppendAuthorized(record);
    this.upsertRecord(record);
    this.recordsMap.set(record.traceId, JSON.stringify(record));
  }

  dataset(
    options: {
      cohortKey?: string | null;
      maxRecords?: number;
    } = {}
  ): GnosisSteeringTraceDataset {
    const indexedRecordIds = this.indexedRecordIds(options.cohortKey);
    const maxRecords = options.maxRecords ?? this.recentWindow;
    const truncated = maxRecords > 0 && indexedRecordIds.length > maxRecords;
    const selectedRecordIds =
      maxRecords > 0 ? indexedRecordIds.slice(-maxRecords) : indexedRecordIds;

    return {
      records: selectedRecordIds
        .map((recordId) => this.recordsById.get(recordId))
        .filter(
          (record): record is GnosisSteeringTraceRecord => record !== undefined
        ),
      invalidLineCount: this.invalidEventCount,
      truncated,
      recentWindow: maxRecords > 0 ? maxRecords : null,
      cohortKey: options.cohortKey ?? null,
    };
  }

  records(
    options: {
      cohortKey?: string | null;
      maxRecords?: number;
    } = {}
  ): GnosisSteeringTraceRecord[] {
    return this.dataset(options).records;
  }

  summarize(
    options: {
      cohortKey?: string | null;
      maxRecords?: number;
    } = {}
  ): GnosisSteeringTraceSummary {
    const dataset = this.dataset(options);

    return summarizeSteeringTraceRecords(dataset.records, {
      cohortKey: dataset.cohortKey,
      invalidLineCount: dataset.invalidLineCount,
      truncated: dataset.truncated,
      recentWindow: dataset.recentWindow,
    });
  }

  private indexedRecordIds(cohortKey?: string | null): string[] {
    if (!cohortKey) {
      return this.recordOrder;
    }

    return this.cohortRecordOrder.get(cohortKey) ?? [];
  }

  private upsertRecord(record: GnosisSteeringTraceRecord): void {
    const previousRecord = this.recordsById.get(record.traceId);
    if (previousRecord) {
      this.removeFromIndexes(previousRecord);
    }

    this.recordsById.set(record.traceId, record);
    this.insertIntoIndexes(record);
  }

  private deleteRecord(traceId: string): void {
    const previousRecord = this.recordsById.get(traceId);
    if (!previousRecord) {
      return;
    }

    this.removeFromIndexes(previousRecord);
    this.recordsById.delete(traceId);
  }

  private insertIntoIndexes(record: GnosisSteeringTraceRecord): void {
    this.insertRecordId(this.recordOrder, record);
    const cohortIndex = this.cohortRecordOrder.get(record.cohortKey) ?? [];
    this.insertRecordId(cohortIndex, record);
    this.cohortRecordOrder.set(record.cohortKey, cohortIndex);
  }

  private removeFromIndexes(record: GnosisSteeringTraceRecord): void {
    this.removeRecordId(this.recordOrder, record.traceId);
    const cohortIndex = this.cohortRecordOrder.get(record.cohortKey);
    if (!cohortIndex) {
      return;
    }

    this.removeRecordId(cohortIndex, record.traceId);
    if (cohortIndex.length === 0) {
      this.cohortRecordOrder.delete(record.cohortKey);
    }
  }

  private insertRecordId(
    recordIds: string[],
    record: GnosisSteeringTraceRecord
  ): void {
    const existingIndex = recordIds.indexOf(record.traceId);
    if (existingIndex >= 0) {
      recordIds.splice(existingIndex, 1);
    }

    let lowerBound = 0;
    let upperBound = recordIds.length;

    while (lowerBound < upperBound) {
      const middle = Math.floor((lowerBound + upperBound) / 2);
      const middleRecordId = recordIds[middle];
      const middleRecord =
        middleRecordId === undefined
          ? undefined
          : this.recordsById.get(middleRecordId);
      if (!middleRecord) {
        lowerBound = middle + 1;
        continue;
      }

      if (compareTraceRecordOrder(middleRecord, record) <= 0) {
        lowerBound = middle + 1;
      } else {
        upperBound = middle;
      }
    }

    recordIds.splice(lowerBound, 0, record.traceId);
  }

  private removeRecordId(recordIds: string[], traceId: string): void {
    const recordIndex = recordIds.indexOf(traceId);
    if (recordIndex >= 0) {
      recordIds.splice(recordIndex, 1);
    }
  }

  private assertRelayConnectAuthorized(): void {
    if (
      !this.relay ||
      !this.executionAuth ||
      this.executionAuth.enforce !== true
    ) {
      return;
    }

    const authorization = authorizeSteeringRelayConnect({
      roomName: this.relay.status.roomName,
      auth: this.executionAuth,
    });
    this.throwIfUnauthorized(authorization, 'steering relay connect denied');
  }

  private assertTraceAppendAuthorized(record: GnosisSteeringTraceRecord): void {
    if (!this.executionAuth || this.executionAuth.enforce !== true) {
      return;
    }

    const authorization = authorizeSteeringTraceAppend({
      cohortKey: record.cohortKey,
      auth: this.executionAuth,
    });
    this.throwIfUnauthorized(
      authorization,
      `steering trace append denied for cohort ${record.cohortKey}`
    );
  }

  private throwIfUnauthorized(
    authorization: GnosisSteeringAuthorizationResult,
    message: string
  ): void {
    if (!authorization.allowed) {
      throw new Error(
        `${message}: ${authorization.reason ?? 'missing capability'}`
      );
    }
  }

  get status(): GnosisSteeringTraceStoreStatus {
    return {
      guid: this.doc.guid,
      durable: this.relay?.status.connected ?? false,
      relayConfigured: this.relay !== null,
      relay: this.relay ? this.relay.status : null,
      localRecordCount: this.recordsById.size,
      cohortCount: this.cohortRecordOrder.size,
      invalidEventCount: this.invalidEventCount,
      lastRelayError: this.lastRelayError,
      authEnforced: this.executionAuth?.enforce === true,
      principal: this.executionAuth?.principal ?? null,
    };
  }
}
