// @a0n/gnosis/src/study-sms-templates.ts
/**
 * Short SMS bodies for study invites/reminders (Twilio transport in apps/edgework-app).
 * Keep under ~300 chars when possible for single-segment GSM-7; callers may truncate URLs.
 */

export interface StudyInviteSmsInput {
  readonly studyTitle: string;
  readonly studyId: string;
  readonly actionUrl: string;
  readonly consentVersion?: string;
}

export interface StudyReminderSmsInput {
  readonly studyTitle: string;
  readonly studyId: string;
  readonly actionUrl: string;
}

function truncate(s: string, max: number): string {
  const t = s.trim();
  return t.length <= max ? t : `${t.slice(0, max - 1)}…`;
}

/** Single-segment-friendly invite text. */
export function renderStudyInviteSms(input: StudyInviteSmsInput): string {
  const title = truncate(input.studyTitle, 72);
  const cv = input.consentVersion?.trim();
  const tail = cv ? ` Consent: ${cv}.` : '';
  return `Study invite: ${title} (id ${input.studyId}).${tail} Link: ${input.actionUrl}`;
}

/**
 * Handles the gnosis render Study Reminder Sms workflow.
 */
export function renderStudyReminderSms(input: StudyReminderSmsInput): string {
  const title = truncate(input.studyTitle, 80);
  return `Reminder: ${title} [${input.studyId}] ${input.actionUrl}`;
}
