// @a0n/gnosis/src/study-participant-link.ts
/**
 * Cross-plane metadata to correlate recruitment / responses with StudyGraphSnapshotV1
 * without placing PII inside the public snapshot JSON.
 *
 * See docs/STUDY_PARTICIPANT_DATA_POLICY.md
 */

import type { StudyGraphSnapshotV1 } from './study-graph-snapshot';

/** Stable ingress envelope for Dash / Knotchain sidecars (no raw PII). */
export interface StudyParticipantIngressRefV1 {
  readonly schemaVersion: '1.0.0';
  /** Same identifier as snapshot.study_id */
  readonly study_id: string;
  /** Semantic consent / IRB packet version (operator-defined string). */
  readonly consent_version: string;
  /** Instrument or form definition version. */
  readonly instrument_version: string;
  /**
   * Hex or base64url hash of opaque invite token + server pepper (operator-defined).
   * Never put raw tokens in snapshot or public logs.
   */
  readonly invite_token_hash?: string;
  /** Optional link to preregistration (may mirror snapshot.preregistration_ref). */
  readonly preregistration_ref?: string;
}

/**
 * Builds the Study Participant Ingress Ref.
 */
export function buildStudyParticipantIngressRef(input: {
  readonly snapshot: Pick<
    StudyGraphSnapshotV1,
    'study_id' | 'preregistration_ref'
  >;
  readonly consentVersion: string;
  readonly instrumentVersion: string;
  readonly inviteTokenHash?: string;
}): StudyParticipantIngressRefV1 {
  return {
    schemaVersion: '1.0.0',
    study_id: input.snapshot.study_id,
    consent_version: input.consentVersion.trim(),
    instrument_version: input.instrumentVersion.trim(),
    invite_token_hash: input.inviteTokenHash,
    preregistration_ref: input.snapshot.preregistration_ref,
  };
}

/** Non-PII aggregate for Knotchain/Dash dashboards (optional sidecar to snapshot). */
export interface StudyAggregateSidecarV1 {
  readonly schemaVersion: '1.0.0';
  readonly study_id: string;
  readonly instrument_version: string;
  readonly response_count: number;
  readonly completed_count: number;
  readonly exported_at: string;
}

/**
 * Builds the Study Aggregate Sidecar.
 */
export function buildStudyAggregateSidecar(input: {
  readonly studyId: string;
  readonly instrumentVersion: string;
  readonly responseCount: number;
  readonly completedCount: number;
  readonly exportedAtIso?: string;
}): StudyAggregateSidecarV1 {
  return {
    schemaVersion: '1.0.0',
    study_id: input.studyId.trim(),
    instrument_version: input.instrumentVersion.trim(),
    response_count: Math.max(0, Math.floor(input.responseCount)),
    completed_count: Math.max(0, Math.floor(input.completedCount)),
    exported_at: input.exportedAtIso ?? new Date().toISOString(),
  };
}
