// @a0n/gnosis/src/tauri-edge-runtime.ts
import type {
  AeonReactorPolicy,
  AeonReactorPolicyRule,
} from './auth/reactor-attestation';
import { hashAeonReactorPolicy } from './auth/reactor-attestation';
import {
  checkGrantedCapability,
  type GnosisCapability,
  type GnosisExecutionAuthContext,
} from './auth/core';
import {
  GnosisWasmExecutionSurface,
  type ExecuteWasmArtifactRequest,
  type GnosisWasmExecutionSurfaceOptions,
  type WasmExecutionResponse,
} from './wasm-execution';
import type { WasmExecutionAbi } from './wasm-execution-bridge';
import {
  type TauriEdgeArtifactManifest,
  type TauriEdgeCapabilityDisposition,
  type TauriEdgeCapabilityKey,
  type TauriEdgeCapabilityPolicyEntry,
  type TauriEdgeZkPolicy,
} from './tauri-edge';

export interface TauriEdgeZkProofEnvelope {
  readonly proof: string;
  readonly paths?: readonly string[];
}

export interface TauriEdgeCommandDescriptor {
  readonly name: string;
  readonly exportName: string;
  readonly abi: WasmExecutionAbi;
}

export interface TauriEdgeCommandSurfaceDocument {
  readonly commands: Record<
    string,
    string | Partial<Omit<TauriEdgeCommandDescriptor, 'name'>>
  >;
}

export interface TauriEdgeInvokeRequest {
  readonly manifest: TauriEdgeArtifactManifest;
  readonly guestWasmPath?: string;
  readonly guestWasmBase64?: string;
  readonly command: string;
  readonly input?: unknown;
  readonly commandSurface?: unknown;
  readonly executionAuth?: GnosisExecutionAuthContext | null;
  readonly runtimeBindingHash?: string | null;
  readonly zkProof?: TauriEdgeZkProofEnvelope | null;
  readonly scopePath?: string;
}

export interface TauriEdgeInvokeResponse {
  readonly status: 'executed' | 'delegated';
  readonly artifactId: string;
  readonly command: string;
  readonly exportName?: string;
  readonly abi?: WasmExecutionAbi;
  readonly payload?: unknown;
  readonly logs?: string;
  readonly executionAuth?: WasmExecutionResponse['executionAuth'];
  readonly capabilityDisposition: TauriEdgeCapabilityDisposition;
  readonly reactorPolicyHash: string;
  readonly runtimeBindingHash?: string;
}

export interface GnosisTauriEdgeExecutionSurfaceOptions {
  readonly wasmSurface?: GnosisWasmExecutionSurface;
  readonly wasmExecution?: GnosisWasmExecutionSurfaceOptions;
  readonly runtimeName?: string;
}

function normalizeOptionalString(
  value: string | null | undefined
): string | undefined {
  const trimmed = value?.trim();
  return trimmed && trimmed.length > 0 ? trimmed : undefined;
}

function requiredCapability(can: string, artifactId: string): GnosisCapability {
  return {
    can,
    with: `aeon://artifact/${artifactId}`,
  };
}

function enforceCapability(
  auth: GnosisExecutionAuthContext | null | undefined,
  required: GnosisCapability,
  reason: string
): void {
  if (!auth || auth.enforce !== true) {
    return;
  }

  if (!checkGrantedCapability(auth.capabilities, required)) {
    throw new Error(
      `Missing capability ${required.can} on ${required.with} for ${reason}.`
    );
  }
}

function resolveCapabilityPolicyEntry(
  manifest: TauriEdgeArtifactManifest,
  key: TauriEdgeCapabilityKey
): TauriEdgeCapabilityPolicyEntry {
  return (
    manifest.capabilityPolicy.find((entry) => entry.key === key) ?? {
      key,
      disposition: 'denied',
      notes: `Capability ${key} is not admitted by this tauri-edge artifact.`,
    }
  );
}

function enforceCapabilityDisposition(
  entry: TauriEdgeCapabilityPolicyEntry,
  action: string
): void {
  if (entry.disposition === 'denied') {
    throw new Error(entry.notes ?? `Tauri-edge ${action} is denied by policy.`);
  }
}

function validateRuntimeBinding(
  manifest: TauriEdgeArtifactManifest,
  runtimeBindingHash: string | null | undefined
): string | undefined {
  const expected = normalizeOptionalString(manifest.runtimeBindingHash);
  const actual = normalizeOptionalString(runtimeBindingHash ?? undefined);

  if (expected && actual && expected !== actual) {
    throw new Error(
      `Tauri-edge runtime binding mismatch: manifest expects ${expected}, received ${actual}.`
    );
  }

  return expected ?? actual;
}

function pathRequiresZk(
  policy: TauriEdgeZkPolicy | undefined,
  scopePath: string | undefined
): string | null {
  if (!policy || !scopePath) {
    return null;
  }

  return (
    policy.protectedPaths.find((candidate) => {
      const normalized = normalizeOptionalString(candidate);
      return normalized ? scopePath.startsWith(normalized) : false;
    }) ?? null
  );
}

function enforceZkPolicy(input: {
  readonly manifest: TauriEdgeArtifactManifest;
  readonly scopePath?: string;
  readonly zkProof?: TauriEdgeZkProofEnvelope | null;
}): void {
  const protectedPath = pathRequiresZk(
    input.manifest.zkPolicy,
    input.scopePath
  );
  if (!protectedPath || !input.manifest.zkPolicy) {
    return;
  }

  if (input.manifest.zkPolicy.mode === 'off') {
    return;
  }

  const proof = input.zkProof;
  const coversPath =
    !!proof?.proof &&
    (proof.paths?.some((candidate) => candidate === protectedPath) ?? false);

  if (input.manifest.zkPolicy.mode === 'required' && !coversPath) {
    throw new Error(
      `Tauri-edge path ${protectedPath} requires a ZK proof under ${
        input.manifest.zkPolicy.proofSystem ?? 'the configured proof system'
      }.`
    );
  }
}

function parseCommandSurfaceDocument(
  source: unknown
): TauriEdgeCommandSurfaceDocument | null {
  if (source === null || source === undefined) {
    return null;
  }

  const normalized = (() => {
    if (source instanceof Uint8Array) {
      return new TextDecoder('utf-8', { fatal: true }).decode(source);
    }
    if (typeof source === 'string') {
      return source;
    }
    return source;
  })();

  const parsed =
    typeof normalized === 'string'
      ? (JSON.parse(normalized) as unknown)
      : normalized;

  if (
    typeof parsed !== 'object' ||
    parsed === null ||
    typeof (parsed as { commands?: unknown }).commands !== 'object' ||
    (parsed as { commands?: unknown }).commands === null
  ) {
    return null;
  }

  return parsed as TauriEdgeCommandSurfaceDocument;
}

function resolveCommandDescriptor(
  manifest: TauriEdgeArtifactManifest,
  command: string,
  commandSurface: unknown
): TauriEdgeCommandDescriptor {
  const normalizedCommand = normalizeOptionalString(command);
  if (!normalizedCommand) {
    throw new Error('Tauri-edge invoke requires a non-empty command name.');
  }

  const document = parseCommandSurfaceDocument(commandSurface);
  const entry = document?.commands?.[normalizedCommand];
  if (typeof entry === 'string') {
    return {
      name: normalizedCommand,
      exportName: entry,
      abi: 'json-v1',
    };
  }

  if (typeof entry === 'object' && entry !== null) {
    const record = entry as Partial<Omit<TauriEdgeCommandDescriptor, 'name'>>;
    return {
      name: normalizedCommand,
      exportName:
        normalizeOptionalString(record.exportName) ?? normalizedCommand,
      abi: record.abi ?? 'json-v1',
    };
  }

  return {
    name: normalizedCommand,
    exportName: normalizedCommand,
    abi: manifest.entrypoints.commandSurface ? 'json-v1' : 'scalar-v1',
  };
}

function buildCapabilityRules(
  manifest: TauriEdgeArtifactManifest
): AeonReactorPolicyRule[] {
  const resource = `aeon://artifact/${manifest.artifactId}`;
  const rules: AeonReactorPolicyRule[] = [
    {
      operation: 'request.read',
      effect: 'allow',
      ucanCapability: 'tauri-edge/invoke',
      resource,
    },
    {
      operation: 'request.metadata',
      effect: 'allow',
      ucanCapability: 'tauri-edge/invoke',
      resource,
    },
    {
      operation: 'response.start',
      effect: 'allow',
      ucanCapability: 'tauri-edge/invoke',
      resource,
    },
    {
      operation: 'response.write',
      effect: 'allow',
      ucanCapability: 'tauri-edge/invoke',
      resource,
    },
    {
      operation: 'response.end',
      effect: 'allow',
      ucanCapability: 'tauri-edge/invoke',
      resource,
    },
  ];

  const materializeRule = (operation: AeonReactorPolicyRule['operation']) => ({
    operation,
    effect: 'allow' as const,
    ucanCapability: 'tauri-edge/materialize',
    resource,
  });

  const isSupported = (key: TauriEdgeCapabilityKey): boolean =>
    resolveCapabilityPolicyEntry(manifest, key).disposition === 'supported';

  if (isSupported('events') || isSupported('http')) {
    rules.push(
      materializeRule('stream.open'),
      materializeRule('stream.read'),
      materializeRule('stream.write'),
      materializeRule('stream.close')
    );
  }

  if (isSupported('fs') || isSupported('storage')) {
    rules.push(
      materializeRule('vfs.open'),
      materializeRule('vfs.read'),
      materializeRule('vfs.write')
    );
  }

  return rules;
}

/**
 * Builds the Tauri Edge Reactor Policy.
 */
export function buildTauriEdgeReactorPolicy(input: {
  readonly manifest: TauriEdgeArtifactManifest;
  readonly runtimeName?: string;
}): AeonReactorPolicy {
  return {
    version: 'aeon_reactor_v1',
    runtime: input.runtimeName ?? 'tauri-edge',
    defaultEffect: 'deny',
    rules: buildCapabilityRules(input.manifest),
  };
}

/**
 * Returns whether hash Tauri Edge Reactor Policy is true.
 */
export function hashTauriEdgeReactorPolicy(input: {
  readonly manifest: TauriEdgeArtifactManifest;
  readonly runtimeName?: string;
}): Promise<string> {
  return hashAeonReactorPolicy(buildTauriEdgeReactorPolicy(input));
}

export class GnosisTauriEdgeExecutionSurface {
  private readonly wasmSurface: GnosisWasmExecutionSurface;
  private readonly runtimeName: string;

  constructor(options: GnosisTauriEdgeExecutionSurfaceOptions = {}) {
    this.wasmSurface =
      options.wasmSurface ??
      new GnosisWasmExecutionSurface(options.wasmExecution);
    this.runtimeName = options.runtimeName ?? 'tauri-edge';
  }

  getWasmSurface(): GnosisWasmExecutionSurface {
    return this.wasmSurface;
  }

  buildReactorPolicy(manifest: TauriEdgeArtifactManifest): AeonReactorPolicy {
    return buildTauriEdgeReactorPolicy({
      manifest,
      runtimeName: this.runtimeName,
    });
  }

  async invoke(
    request: TauriEdgeInvokeRequest
  ): Promise<TauriEdgeInvokeResponse> {
    const invokePolicy = resolveCapabilityPolicyEntry(
      request.manifest,
      'invoke'
    );
    enforceCapabilityDisposition(invokePolicy, 'invoke');
    const runtimeBindingHash = validateRuntimeBinding(
      request.manifest,
      request.runtimeBindingHash
    );
    enforceZkPolicy({
      manifest: request.manifest,
      ...(request.scopePath !== undefined
        ? { scopePath: request.scopePath }
        : {}),
      ...(request.zkProof !== undefined ? { zkProof: request.zkProof } : {}),
    });
    enforceCapability(
      request.executionAuth,
      requiredCapability('tauri-edge/invoke', request.manifest.artifactId),
      'invoke'
    );

    if (invokePolicy.disposition === 'delegated') {
      return {
        status: 'delegated',
        artifactId: request.manifest.artifactId,
        command: request.command,
        capabilityDisposition: invokePolicy.disposition,
        reactorPolicyHash: await hashAeonReactorPolicy(
          this.buildReactorPolicy(request.manifest)
        ),
        ...(runtimeBindingHash ? { runtimeBindingHash } : {}),
      };
    }

    const guestWasmPath = normalizeOptionalString(request.guestWasmPath);
    const guestWasmBase64 = normalizeOptionalString(request.guestWasmBase64);
    if (!guestWasmPath && !guestWasmBase64) {
      throw new Error(
        `Tauri-edge invoke for ${request.manifest.artifactId} requires a guest Wasm payload.`
      );
    }

    const command = resolveCommandDescriptor(
      request.manifest,
      request.command,
      request.commandSurface
    );
    const registeredArtifact = this.wasmSurface.register({
      artifactId: `${request.manifest.artifactId}:${command.exportName}`,
      ...(guestWasmPath ? { wasmPath: guestWasmPath } : {}),
      ...(guestWasmBase64 ? { wasmBase64: guestWasmBase64 } : {}),
      exportName: command.exportName,
      abi: command.abi,
    });
    const wasmExecutionAuth =
      request.executionAuth && request.executionAuth.enforce === true
        ? {
            ...request.executionAuth,
            enforce: false,
          }
        : request.executionAuth;
    const wasmRequest: ExecuteWasmArtifactRequest = {
      artifactId: registeredArtifact.artifactId,
      input: request.input,
      ...(wasmExecutionAuth
        ? { executionAuth: wasmExecutionAuth }
        : {}),
    };
    const response = await this.wasmSurface.execute(wasmRequest);
    return {
      status: 'executed',
      artifactId: request.manifest.artifactId,
      command: command.name,
      exportName: command.exportName,
      abi: command.abi,
      payload: response.payload,
      logs: response.logs,
      executionAuth: response.executionAuth,
      capabilityDisposition: invokePolicy.disposition,
      reactorPolicyHash: await hashAeonReactorPolicy(
        this.buildReactorPolicy(request.manifest)
      ),
      ...(runtimeBindingHash ? { runtimeBindingHash } : {}),
    };
  }
}
