// @a0n/gnosis/src/tauri-edge.ts
import { createHash } from 'node:crypto';
import { XGnosisEdgeVM } from './x-gnosis-vm';

export const TAURI_EDGE_FLOW_HEADER_SIZE = 10;
export const TAURI_EDGE_FLOW_FIN = 0x01;
export const DEFAULT_TAURI_EDGE_FRAGMENT_SIZE_BYTES = 64 * 1024;

export type TauriEdgeFragmentFamily =
  | 'frontend'
  | 'guest-wasm'
  | 'command-surface'
  | 'assets'
  | 'snapshot'
  | 'attestation'
  | 'capability';

export type TauriEdgeCapabilityKey =
  | 'invoke'
  | 'events'
  | 'http'
  | 'storage'
  | 'fs'
  | 'shell'
  | 'updater'
  | 'tray'
  | 'global-shortcut'
  | 'process'
  | 'plugin';

export type TauriEdgeCapabilityDisposition =
  | 'supported'
  | 'delegated'
  | 'denied';

export interface TauriEdgeFlowFrame {
  readonly streamId: number;
  readonly sequence: number;
  readonly flags: number;
  readonly payload: Uint8Array;
}

export interface TauriEdgeCapabilityPolicyEntry {
  readonly key: TauriEdgeCapabilityKey;
  readonly disposition: TauriEdgeCapabilityDisposition;
  readonly notes?: string;
}

export interface TauriEdgeZkPolicy {
  readonly mode: 'off' | 'preferred' | 'required';
  readonly protectedPaths: readonly string[];
  readonly proofSystem?: string;
}

export interface TauriEdgeFragmentDescriptor {
  readonly streamId: number;
  readonly ordinal: number;
  readonly family: TauriEdgeFragmentFamily;
  readonly encoding: 'flowframe-v1';
  readonly byteLength: number;
  readonly payloadByteLength: number;
  readonly hash: string;
  readonly payloadRef: string;
  readonly sourcePath?: string;
  readonly contentType?: string;
}

export interface TauriEdgeArtifactManifest {
  readonly version: 1;
  readonly runtimeTarget: 'tauri-edge';
  readonly artifactId: string;
  readonly application: {
    readonly name: string;
    readonly identifier?: string;
    readonly productName?: string;
    readonly version: string;
    readonly configPath: string;
    readonly frontendRoot?: string;
  };
  readonly bundleDigest: string;
  readonly manifestDigest: string;
  readonly frontendDigest?: string;
  readonly guestDigest?: string;
  readonly capabilityDigest: string;
  readonly attestationDigest?: string;
  readonly commandSurfaceDigest?: string;
  readonly ucanAudience: string;
  readonly ucanRequiredCaps: readonly string[];
  readonly runtimeBindingHash?: string;
  readonly snapshotCompatibility: 'reconstructable-v1';
  readonly entrypoints: {
    readonly frontend?: string;
    readonly guestWasm?: string;
    readonly commandSurface?: string;
    readonly capabilityManifest?: string;
  };
  readonly capabilityPolicy: readonly TauriEdgeCapabilityPolicyEntry[];
  readonly zkPolicy?: TauriEdgeZkPolicy;
  readonly fragments: readonly TauriEdgeFragmentDescriptor[];
  readonly createdAt: string;
}

export interface TauriEdgeD1TableNames {
  readonly artifacts: string;
  readonly fragments: string;
  readonly sessions: string;
}

export interface TauriEdgeArtifactRow {
  readonly artifact_id: string;
  readonly version: number;
  readonly runtime_target: 'tauri-edge';
  readonly manifest_digest: string;
  readonly bundle_digest: string;
  readonly frontend_digest?: string;
  readonly guest_digest?: string;
  readonly capability_digest: string;
  readonly attestation_digest?: string;
  readonly command_surface_digest?: string;
  readonly ucan_audience: string;
  readonly ucan_required_caps: string;
  readonly runtime_binding_hash?: string;
  readonly snapshot_compatibility: 'reconstructable-v1';
  readonly created_at: string;
  readonly manifest_json: string;
}

export interface TauriEdgeFragmentRow {
  readonly artifact_id: string;
  readonly stream_id: number;
  readonly ordinal: number;
  readonly family: TauriEdgeFragmentFamily;
  readonly encoding: 'flowframe-v1';
  readonly byte_length: number;
  readonly payload_byte_length: number;
  readonly hash: string;
  readonly payload_ref: string;
  readonly source_path?: string;
  readonly content_type?: string;
}

export interface TauriEdgeSessionRow {
  readonly session_id: string;
  readonly artifact_id: string;
  readonly snapshot_root?: string;
  readonly position_root?: string;
  readonly stream_id?: number;
  readonly ordinal?: number;
  readonly zk_proof?: string;
  readonly updated_at: number;
}

function stableStringify(value: unknown): string {
  if (value === null || value === undefined) {
    return 'null';
  }

  if (typeof value === 'string') {
    return JSON.stringify(value);
  }

  if (typeof value === 'number') {
    return Number.isFinite(value) ? String(value) : 'null';
  }

  if (typeof value === 'boolean') {
    return value ? 'true' : 'false';
  }

  if (typeof value === 'bigint') {
    return JSON.stringify(value.toString(10));
  }

  if (value instanceof Uint8Array) {
    return `[${Array.from(value).join(',')}]`;
  }

  if (Array.isArray(value)) {
    return `[${value.map((entry) => stableStringify(entry)).join(',')}]`;
  }

  if (typeof value === 'object') {
    const entries = Object.entries(value as Record<string, unknown>)
      .filter(([, entry]) => entry !== undefined)
      .sort(([left], [right]) => left.localeCompare(right));

    return `{${entries
      .map(([key, entry]) => `${JSON.stringify(key)}:${stableStringify(entry)}`)
      .join(',')}}`;
  }

  return 'null';
}

/**
 * Handles the gnosis sha256 Hex workflow.
 */
export function sha256Hex(value: Uint8Array | string): string {
  return createHash('sha256').update(value).digest('hex');
}

/**
 * Handles the gnosis encode Tauri Edge Flow Frame workflow.
 */
export function encodeTauriEdgeFlowFrame(
  frame: TauriEdgeFlowFrame
): Uint8Array {
  const bytes = new Uint8Array(
    TAURI_EDGE_FLOW_HEADER_SIZE + frame.payload.length
  );
  const view = new DataView(bytes.buffer);

  view.setUint16(0, frame.streamId, false);
  view.setUint32(2, frame.sequence >>> 0, false);
  view.setUint8(6, frame.flags & 0xff);
  bytes[7] = (frame.payload.length >>> 16) & 0xff;
  bytes[8] = (frame.payload.length >>> 8) & 0xff;
  bytes[9] = frame.payload.length & 0xff;
  bytes.set(frame.payload, TAURI_EDGE_FLOW_HEADER_SIZE);
  return bytes;
}

/**
 * Handles the gnosis decode Tauri Edge Flow Frame workflow.
 */
export function decodeTauriEdgeFlowFrame(
  bytes: Uint8Array,
  vmLimits?: { budgetR: number; initialV: number }
): TauriEdgeFlowFrame {
  if (bytes.length < TAURI_EDGE_FLOW_HEADER_SIZE) {
    throw new Error(
      'Tauri-edge fragment frame is shorter than the 10-byte header.'
    );
  }

  const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  const byte7 = bytes[7];
  const byte8 = bytes[8];
  const byte9 = bytes[9];
  if (byte7 === undefined || byte8 === undefined || byte9 === undefined) {
    throw new Error('Tauri-edge fragment frame is missing header bytes.');
  }
  const payloadLength = (byte7 << 16) | (byte8 << 8) | byte9;
  const expectedLength = TAURI_EDGE_FLOW_HEADER_SIZE + payloadLength;
  if (bytes.length !== expectedLength) {
    throw new Error(
      `Tauri-edge fragment frame length mismatch: expected ${expectedLength} bytes, received ${bytes.length}.`
    );
  }

  const frame = {
    streamId: view.getUint16(0, false),
    sequence: view.getUint32(2, false),
    flags: view.getUint8(6),
    payload: bytes.slice(TAURI_EDGE_FLOW_HEADER_SIZE),
  };

  // processFrame treats payload byte length as frame complexity; budget must scale with payload
  // so large artifact fragments (e.g. WASM) decode without Stress-Energy Collapse.
  const budgetR =
    vmLimits?.budgetR ?? Math.max(54, frame.payload.length);
  const initialV = vmLimits?.initialV ?? 0;
  const vm = new XGnosisEdgeVM(budgetR, initialV);
  // Universal Law: Thermodynamic Mathematical constraints apply rigidly to ALL cross-platform payloads
  vm.processFrame(frame);

  return frame;
}

/**
 * Builds the Tauri Edge Capability Grant.
 */
export function buildTauriEdgeCapabilityGrant(
  can: string,
  resource: string
): string {
  return `${can} ${resource}`;
}

/**
 * Builds the Tauri Edge Required Ucan Capabilities.
 */
export function buildTauriEdgeRequiredUcanCapabilities(
  artifactId: string
): string[] {
  const resource = `aeon://artifact/${artifactId}`;
  return [
    buildTauriEdgeCapabilityGrant('tauri-edge/artifact.hydrate', resource),
    buildTauriEdgeCapabilityGrant('tauri-edge/fragment.read', resource),
    buildTauriEdgeCapabilityGrant('tauri-edge/session.resume', resource),
    buildTauriEdgeCapabilityGrant('tauri-edge/invoke', resource),
    buildTauriEdgeCapabilityGrant('tauri-edge/materialize', resource),
  ];
}

/**
 * Builds the Default Tauri Edge Capability Policy.
 */
export function buildDefaultTauriEdgeCapabilityPolicy(input?: {
  readonly hasGuestWasm?: boolean;
}): TauriEdgeCapabilityPolicyEntry[] {
  const hasGuestWasm = input?.hasGuestWasm ?? false;
  return [
    {
      key: 'invoke',
      disposition: hasGuestWasm ? 'supported' : 'delegated',
      ...(hasGuestWasm
        ? {}
        : {
            notes:
              'Guest Wasm was not provided, so command invocation must be delegated or rejected upstream.',
          }),
    },
    { key: 'events', disposition: 'supported' },
    { key: 'http', disposition: 'supported' },
    { key: 'storage', disposition: 'supported' },
    { key: 'fs', disposition: 'supported' },
    {
      key: 'shell',
      disposition: 'delegated',
      notes:
        'Native shell access is not available inside the semantic worker runtime.',
    },
    {
      key: 'updater',
      disposition: 'denied',
      notes: 'Native updater flows stay on the desktop/mobile bundle lane.',
    },
    {
      key: 'tray',
      disposition: 'denied',
      notes: 'System tray integration is a native-only surface.',
    },
    {
      key: 'global-shortcut',
      disposition: 'denied',
      notes: 'Global shortcuts are a native-only surface.',
    },
    {
      key: 'process',
      disposition: 'denied',
      notes:
        'Raw process spawn is not admitted by the tauri-edge worker runtime.',
    },
    {
      key: 'plugin',
      disposition: 'delegated',
      notes:
        'Native plugin ABI must be delegated or replaced by a kernel capability.',
    },
  ];
}

/**
 * Resolves the Tauri Edge D1 Table Names.
 */
export function resolveTauriEdgeD1TableNames(
  prefix = 'tauri_edge'
): TauriEdgeD1TableNames {
  const normalized =
    prefix.trim().replace(/[^A-Za-z0-9_]+/g, '_') || 'tauri_edge';
  return {
    artifacts: `${normalized}_artifacts`,
    fragments: `${normalized}_fragments`,
    sessions: `${normalized}_sessions`,
  };
}

/**
 * Builds the Tauri Edge D1 Schema.
 */
export function buildTauriEdgeD1Schema(
  tables: TauriEdgeD1TableNames = resolveTauriEdgeD1TableNames()
): string[] {
  return [
    `
      CREATE TABLE IF NOT EXISTS ${tables.artifacts} (
        artifact_id TEXT PRIMARY KEY,
        version INTEGER NOT NULL,
        runtime_target TEXT NOT NULL,
        manifest_digest TEXT NOT NULL,
        bundle_digest TEXT NOT NULL,
        frontend_digest TEXT,
        guest_digest TEXT,
        capability_digest TEXT NOT NULL,
        attestation_digest TEXT,
        command_surface_digest TEXT,
        ucan_audience TEXT NOT NULL,
        ucan_required_caps TEXT NOT NULL,
        runtime_binding_hash TEXT,
        snapshot_compatibility TEXT NOT NULL,
        created_at TEXT NOT NULL,
        manifest_json TEXT NOT NULL
      )
    `,
    `
      CREATE TABLE IF NOT EXISTS ${tables.fragments} (
        artifact_id TEXT NOT NULL,
        stream_id INTEGER NOT NULL,
        ordinal INTEGER NOT NULL,
        family TEXT NOT NULL,
        encoding TEXT NOT NULL,
        byte_length INTEGER NOT NULL,
        payload_byte_length INTEGER NOT NULL,
        hash TEXT NOT NULL,
        payload_ref TEXT NOT NULL,
        source_path TEXT,
        content_type TEXT,
        PRIMARY KEY (artifact_id, stream_id, ordinal)
      )
    `,
    `
      CREATE TABLE IF NOT EXISTS ${tables.sessions} (
        session_id TEXT PRIMARY KEY,
        artifact_id TEXT NOT NULL,
        snapshot_root TEXT,
        position_root TEXT,
        stream_id INTEGER,
        ordinal INTEGER,
        zk_proof TEXT,
        updated_at INTEGER NOT NULL
      )
    `,
  ].map((statement) => statement.trim());
}

/**
 * Computes the Tauri Edge Manifest Digest.
 */
export function computeTauriEdgeManifestDigest(
  manifest: Omit<TauriEdgeArtifactManifest, 'manifestDigest'>
): string {
  return sha256Hex(stableStringify(manifest));
}

/**
 * Builds the Tauri Edge Artifact Id.
 */
export function buildTauriEdgeArtifactId(input: {
  readonly bundleDigest: string;
  readonly capabilityDigest: string;
  readonly applicationName: string;
  readonly version: string;
}): string {
  return `tauri-edge-${sha256Hex(
    stableStringify({
      name: input.applicationName,
      version: input.version,
      bundleDigest: input.bundleDigest,
      capabilityDigest: input.capabilityDigest,
    })
  ).slice(0, 24)}`;
}

/**
 * Converts input into Tauri Edge Artifact Row.
 */
export function toTauriEdgeArtifactRow(
  manifest: TauriEdgeArtifactManifest
): TauriEdgeArtifactRow {
  return {
    artifact_id: manifest.artifactId,
    version: manifest.version,
    runtime_target: manifest.runtimeTarget,
    manifest_digest: manifest.manifestDigest,
    bundle_digest: manifest.bundleDigest,
    ...(manifest.frontendDigest
      ? { frontend_digest: manifest.frontendDigest }
      : {}),
    ...(manifest.guestDigest ? { guest_digest: manifest.guestDigest } : {}),
    capability_digest: manifest.capabilityDigest,
    ...(manifest.attestationDigest
      ? { attestation_digest: manifest.attestationDigest }
      : {}),
    ...(manifest.commandSurfaceDigest
      ? { command_surface_digest: manifest.commandSurfaceDigest }
      : {}),
    ucan_audience: manifest.ucanAudience,
    ucan_required_caps: JSON.stringify(manifest.ucanRequiredCaps),
    ...(manifest.runtimeBindingHash
      ? { runtime_binding_hash: manifest.runtimeBindingHash }
      : {}),
    snapshot_compatibility: manifest.snapshotCompatibility,
    created_at: manifest.createdAt,
    manifest_json: JSON.stringify(manifest),
  };
}
