// @a0n/gnosis/src/time-spent-waiting.ts
/**
 * time-spent-waiting.ts -- The topology of waiting, anticipation, and fear
 *
 * Waiting is not a passive delay; it is an active computation (a "void walk")
 * where the MonotoneLog advances without the awaited FoldedKnotState resolving.
 * The subjective weight of the wait is governed by the topological delta
 * between the current state and the expected target state, compounded by
 * uncertainty (which operates as Bule -- emotional temperature).
 *
 * Mechanized in TimeSpentWaiting.lean (zero sorry):
 *   - waiting_requires_void_gap: a wait implies a non-zero temporal distance
 *   - fear_is_anticipated_entanglement: expected crossings > current = Fear
 *   - anticipation_is_expected_reidemeister: expected crossings < current = Anticipation
 *   - uncertainty_generates_phantoms: variance in target creates temporary load
 *   - relief_is_phantom_garbage_collection: resolution drops phantom load to zero
 *   - waiting_master_theorem: qualitative experience matches the crossing delta
 *
 * @module @a0n/gnosis/time-spent-waiting
 */

// ============================================================================
// Core Types
// ============================================================================

/** Classification of waiting states by expected topological outcome. */
export type WaitStateKind =
  | 'fear' // Anticipating a more complex knot (trauma)
  | 'anticipation' // Anticipating a simpler knot (Reidemeister move)
  | 'anxiety'; // Unbounded variance, high bule independent of outcome

/** A Wait State -- an observation of a temporal gap. */
export interface WaitState {
  /** Qualitative experience of the wait. */
  readonly kind: WaitStateKind;
  /** The complexity of the present moment. */
  readonly currentCrossings: number;
  /** The forecasted complexity of the expected target event. */
  readonly targetCrossings: number;
  /** The temporal distance waiting to be crossed (ticks of MonotoneLog). */
  readonly voidGap: number;
  /** Variance/Phantom load generated per unit of void gap. Analogous to "Bule" (emotional temperature). */
  readonly uncertaintyBule: number;
  /** Accumulated cognitive weight while waiting (uncertainty * voidGap). */
  readonly phantomLoad: number;
  /** Lean theorem anchor for this state. */
  readonly leanAnchor: string;
}

// ============================================================================
// Factory functions
// ============================================================================

/**
 * Creates a Fear wait state.
 * Per fear_is_anticipated_entanglement: If the target has more crossings
 * than the current state, the projection is fear.
 */
export function fearState(
  currentCrossings: number,
  addedTrauma: number,
  voidGap: number,
  uncertaintyBule: number = 1
): WaitState {
  const targetCrossings = currentCrossings + Math.max(1, addedTrauma);
  return {
    kind: 'fear',
    currentCrossings,
    targetCrossings,
    voidGap,
    uncertaintyBule,
    phantomLoad: Math.max(0, uncertaintyBule * voidGap),
    leanAnchor: 'fear_is_anticipated_entanglement',
  };
}

/**
 * Creates an Anticipation wait state.
 * Per anticipation_is_expected_reidemeister: If the target resolves crossings
 * (a fold), the projection is anticipation.
 */
export function anticipationState(
  currentCrossings: number,
  resolvedCrossings: number,
  voidGap: number,
  uncertaintyBule: number = 1
): WaitState {
  const targetCrossings = Math.max(0, currentCrossings - Math.max(1, resolvedCrossings));
  return {
    kind: 'anticipation',
    currentCrossings,
    targetCrossings,
    voidGap,
    uncertaintyBule,
    phantomLoad: Math.max(0, uncertaintyBule * voidGap),
    leanAnchor: 'anticipation_is_expected_reidemeister',
  };
}

/**
 * Creates an Anxiety wait state.
 * Per anxiety_is_unbounded_variance (anti): High uncertainty/bule dominates
 * the projection regardless of the specific topological delta.
 */
export function anxietyState(
  currentCrossings: number,
  voidGap: number,
  highUncertaintyBule: number = 100
): WaitState {
  return {
    kind: 'anxiety',
    currentCrossings,
    targetCrossings: currentCrossings, // Unbounded variance usually overshadows specific targets
    voidGap,
    uncertaintyBule: highUncertaintyBule,
    phantomLoad: highUncertaintyBule * voidGap,
    leanAnchor: 'anxiety_is_unbounded_variance',
  };
}

// ============================================================================
// State Operations
// ============================================================================

/**
 * Compute the accumulated dread over time.
 * Per dread_compounds_over_time: The longer the clock ticks without resolution,
 * the heavier the phantom load becomes, scaling by the uncertainty (bule).
 */
export function tickWait(state: WaitState, elapsedTicks: number = 1): WaitState {
  const newGap = state.voidGap + elapsedTicks;
  return {
    ...state,
    voidGap: newGap,
    phantomLoad: state.uncertaintyBule * newGap,
  };
}

/**
 * Resolve the wait state.
 * Per relief_is_phantom_garbage_collection: If the event resolves without
 * the anticipated trauma, the accumulated phantom load is immediately discarded.
 * Returns the final stabilized crossing number and 0 phantom load.
 */
export function collapseWait(state: WaitState, actualTargetCrossings: number): {
  crossings: number;
  phantomLoad: number;
  disappointment: boolean;
} {
  // If we anticipated a fold (target < current), but actual > target, it's a failed fold.
  const isDisappointment =
    state.kind === 'anticipation' && actualTargetCrossings > state.targetCrossings;
  
  return {
    crossings: actualTargetCrossings,
    phantomLoad: 0, // relief_is_phantom_garbage_collection
    disappointment: isDisappointment, // disappointment_is_failed_fold
  };
}
