// @a0n/gnosis/src/tla-bridge.ts
import {
  getGgRootNodeIds,
  getGgTerminalNodeIds,
  parseGgProgram,
  type GgEdge,
  type GgProgram,
} from '@a0n/aeon-logic';
import {
  inferCapabilitiesFromGgSource,
  summarizeCapabilityRequirements,
  type CapabilityContractSummary,
} from './capabilities/index';
import { lowerUfcsSource } from './ufcs';

function basenameLike(filePath: string): string {
  const normalized = filePath.replace(/\\/g, '/').replace(/\/+$/, '');
  const segments = normalized.split('/');
  return segments[segments.length - 1] ?? normalized;
}

function extnameLike(filePath: string): string {
  const baseName = basenameLike(filePath);
  const extensionStart = baseName.lastIndexOf('.');
  return extensionStart > 0 ? baseName.slice(extensionStart) : '';
}

export interface GnosisTlaBridgeOptions {
  readonly moduleName?: string;
  readonly sourceFilePath?: string;
}

export interface GnosisTlaBridgeStats {
  readonly nodeCount: number;
  readonly edgeCount: number;
  readonly rootCount: number;
  readonly terminalCount: number;
  readonly foldTargetCount: number;
  readonly forkEdgeCount: number;
  readonly raceEdgeCount: number;
}

export interface GnosisTlaBridgeResult {
  readonly moduleName: string;
  readonly tla: string;
  readonly cfg: string;
  readonly stats: GnosisTlaBridgeStats;
  readonly effects: CapabilityContractSummary;
}

const MODULE_PREFIX = 'Gnosis';
const HEADER_DASHES = '-'.repeat(30);

function sanitizeModuleName(rawName: string): string {
  const normalized = rawName
    .replace(/[^A-Za-z0-9_]/g, '_')
    .replace(/_+/g, '_')
    .replace(/^_+/, '')
    .replace(/_+$/, '');

  if (normalized.length === 0) {
    return `${MODULE_PREFIX}Spec`;
  }

  if (/^[A-Za-z_]/.test(normalized)) {
    return normalized;
  }

  return `${MODULE_PREFIX}_${normalized}`;
}

function deriveModuleName(
  program: GgProgram,
  options: GnosisTlaBridgeOptions
): string {
  if (options.moduleName && options.moduleName.trim().length > 0) {
    return sanitizeModuleName(options.moduleName.trim());
  }

  if (options.sourceFilePath && options.sourceFilePath.trim().length > 0) {
    const extension = extnameLike(options.sourceFilePath);
    const baseName = basenameLike(options.sourceFilePath);
    const moduleSource = extension
      ? baseName.slice(0, Math.max(0, baseName.length - extension.length))
      : baseName;
    return sanitizeModuleName(moduleSource);
  }

  const rootNode = getResolvedRoots(program)[0];
  if (rootNode && rootNode.length > 0) {
    return sanitizeModuleName(`${MODULE_PREFIX}_${rootNode}`);
  }

  return `${MODULE_PREFIX}Spec`;
}

function escapeTlaString(value: string): string {
  return value.replace(/\\/g, '\\\\').replace(/"/g, '\\"');
}

function toTlaSet(nodeIds: readonly string[]): string {
  const uniqueNodeIds = [...new Set(nodeIds)];
  const serialized = uniqueNodeIds
    .map((nodeId) => `"${escapeTlaString(nodeId)}"`)
    .join(', ');
  return `{${serialized}}`;
}

function getResolvedRoots(program: GgProgram): readonly string[] {
  const roots = getGgRootNodeIds(program);
  if (roots.length > 0) {
    return roots;
  }

  const fallbackRoot = program.edges[0]?.sourceIds[0];
  return fallbackRoot ? [fallbackRoot] : [];
}

function getFoldTargets(program: GgProgram): readonly string[] {
  return [
    ...new Set(
      program.edges
        .filter((edge) => {
          const type = edge.type.toUpperCase();
          return type === 'FOLD' || type === 'COLLAPSE' || type === 'OBSERVE';
        })
        .flatMap((edge) => edge.targetIds)
    ),
  ];
}

function isCollapseEdge(edge: GgEdge): boolean {
  const type = edge.type.toUpperCase();
  return type === 'RACE' || type === 'FOLD' || type === 'COLLAPSE';
}

function getCorridorTargets(program: GgProgram): readonly string[] {
  return [
    ...new Set(
      program.edges
        .filter((edge) => {
          if (!isCollapseEdge(edge)) {
            return false;
          }

          const corridor = edge.properties.corridor;
          return typeof corridor === 'string' && corridor.trim().length > 0;
        })
        .flatMap((edge) => edge.targetIds)
    ),
  ];
}

function edgeActionName(edge: GgEdge, index: number): string {
  const typeName = edge.type.toUpperCase().replace(/[^A-Z0-9_]/g, '_');
  const ordinal = String(index + 1).padStart(2, '0');
  return `Edge_${ordinal}_${typeName}`;
}

function edgeBetaExpression(
  edge: GgEdge,
  sourceSet: string,
  targetSet: string
): string {
  const edgeType = edge.type.toUpperCase();
  if (edgeType === 'FORK') {
    return `beta1 + (Cardinality(${targetSet}) - 1)`;
  }
  if (edgeType === 'RACE') {
    return `Max2(0, beta1 - (Cardinality(${sourceSet}) - 1))`;
  }
  if (
    edgeType === 'FOLD' ||
    edgeType === 'COLLAPSE' ||
    edgeType === 'OBSERVE'
  ) {
    return '0';
  }
  if (edgeType === 'VENT' || edgeType === 'TUNNEL') {
    return 'Max2(0, beta1 - 1)';
  }
  return 'beta1';
}

function renderEdgeAction(edge: GgEdge, index: number): string {
  const sourceSet = toTlaSet(edge.sourceIds);
  const targetSet = toTlaSet(edge.targetIds);
  const actionName = edgeActionName(edge, index);
  const betaExpression = edgeBetaExpression(edge, sourceSet, targetSet);
  const edgeType = edge.type.toUpperCase();

  if (edgeType === 'RACE') {
    return [
      `${actionName} ==`,
      `  /\\ CanFire(${sourceSet})`,
      `  /\\ \\E winner \\in ${targetSet}:`,
      `      /\\ active' = UpdateActive(${sourceSet}, {winner})`,
      `      /\\ beta1' = ${betaExpression}`,
      `      /\\ payloadPresent' = payloadPresent`,
      `      /\\ consensusReached' = consensusReached \\/ (winner \\in FOLD_TARGETS)`,
    ].join('\n');
  }

  return [
    `${actionName} ==`,
    `  /\\ CanFire(${sourceSet})`,
    `  /\\ active' = UpdateActive(${sourceSet}, ${targetSet})`,
    `  /\\ beta1' = ${betaExpression}`,
    `  /\\ payloadPresent' = payloadPresent`,
    `  /\\ consensusReached' = consensusReached \\/ (${targetSet} \\cap FOLD_TARGETS # {})`,
  ].join('\n');
}

function renderNext(actionNames: readonly string[]): string {
  if (actionNames.length === 0) {
    return 'Next == FALSE';
  }

  const lines = ['Next =='];
  for (const actionName of actionNames) {
    lines.push(`  \\/ ${actionName}`);
  }
  return lines.join('\n');
}

function renderTla(
  moduleName: string,
  program: GgProgram,
  roots: readonly string[],
  terminals: readonly string[],
  foldTargets: readonly string[],
  corridorTargets: readonly string[],
  effects: CapabilityContractSummary
): string {
  const renderedActions = program.edges.map((edge, index) =>
    renderEdgeAction(edge, index)
  );
  const actionNames = program.edges.map((edge, index) =>
    edgeActionName(edge, index)
  );

  const lines: string[] = [];
  lines.push(`${HEADER_DASHES} MODULE ${moduleName} ${HEADER_DASHES}`);
  lines.push('EXTENDS Naturals, FiniteSets, Sequences');
  lines.push('');
  lines.push(`NODES == ${toTlaSet(program.nodes.map((node) => node.id))}`);
  lines.push(`ROOTS == ${toTlaSet(roots)}`);
  lines.push(`TERMINALS == ${toTlaSet(terminals)}`);
  lines.push(`FOLD_TARGETS == ${toTlaSet(foldTargets)}`);
  lines.push(`CORRIDOR_TARGETS == ${toTlaSet(corridorTargets)}`);
  lines.push(`EFFECTS == ${toTlaSet(effects.required)}`);
  lines.push(`DECLARED_EFFECTS == ${toTlaSet(effects.declared)}`);
  lines.push(`INFERRED_EFFECTS == ${toTlaSet(effects.inferred)}`);
  for (const effectNode of effects.nodes) {
    lines.push(
      `\\* EFFECT_NODE ${effectNode.nodeId} declared=${toTlaSet(
        effectNode.declared
      )} inferred=${toTlaSet(effectNode.inferred)}`
    );
  }
  lines.push('');
  lines.push('VARIABLES active, beta1, payloadPresent, consensusReached');
  lines.push('vars == <<active, beta1, payloadPresent, consensusReached>>');
  lines.push('');
  lines.push('Max2(a, b) == IF a > b THEN a ELSE b');
  lines.push('CanFire(sourceSet) == sourceSet \\subseteq active');
  lines.push(
    'UpdateActive(sourceSet, targetSet) == (active \\ sourceSet) \\cup targetSet'
  );
  lines.push('');
  lines.push('Init ==');
  lines.push('  /\\ active = ROOTS');
  lines.push('  /\\ beta1 = 0');
  lines.push('  /\\ payloadPresent = TRUE');
  lines.push('  /\\ consensusReached = FALSE');
  lines.push('');
  lines.push(...renderedActions);
  lines.push('');
  lines.push(renderNext(actionNames));
  lines.push('');
  lines.push('TypeInvariant ==');
  lines.push('  /\\ active \\subseteq NODES');
  lines.push('  /\\ beta1 \\in Nat');
  lines.push('  /\\ payloadPresent \\in BOOLEAN');
  lines.push('  /\\ consensusReached \\in BOOLEAN');
  lines.push('');
  lines.push('NoLostPayloadInvariant == payloadPresent = TRUE');
  lines.push('HasFoldTargets == FOLD_TARGETS # {}');
  lines.push('HasCorridorTargets == CORRIDOR_TARGETS # {}');
  lines.push('EventuallyTerminal == <> (active \\cap TERMINALS # {})');
  lines.push(
    'EventuallyConsensus == IF HasFoldTargets THEN <> consensusReached ELSE TRUE'
  );
  lines.push(
    'EventuallyMiddleOutCompression == IF HasCorridorTargets THEN <> (active \\cap CORRIDOR_TARGETS # {}) ELSE TRUE'
  );
  lines.push(
    'FirstSufficientCompression == IF HasCorridorTargets THEN <> ((active \\cap CORRIDOR_TARGETS # {}) /\\ beta1 <= Cardinality(CORRIDOR_TARGETS)) ELSE TRUE'
  );
  lines.push('DeadlockFree == []<>(ENABLED Next)');
  lines.push('');
  lines.push('Spec ==');
  lines.push('  /\\ Init');
  lines.push('  /\\ [][Next]_vars');
  lines.push('  /\\ WF_vars(Next)');
  lines.push('');
  lines.push('THEOREM Spec => []NoLostPayloadInvariant');
  lines.push('');
  lines.push(
    '============================================================================='
  );

  return `${lines.join('\n')}\n`;
}

function renderCfg(): string {
  return [
    'SPECIFICATION Spec',
    'INVARIANT TypeInvariant',
    'INVARIANT NoLostPayloadInvariant',
    'PROPERTY DeadlockFree',
    'PROPERTY EventuallyTerminal',
    'PROPERTY EventuallyConsensus',
    'PROPERTY EventuallyMiddleOutCompression',
    'PROPERTY FirstSufficientCompression',
    '',
  ].join('\n');
}

/**
 * Handles the gnosis generate Tla From Gnosis Source workflow.
 */
export function generateTlaFromGnosisSource(
  sourceText: string,
  options: GnosisTlaBridgeOptions = {}
): GnosisTlaBridgeResult {
  const normalizedSource = lowerUfcsSource(sourceText);
  const program = parseGgProgram(normalizedSource);
  const moduleName = deriveModuleName(program, options);
  const roots = getResolvedRoots(program);
  const terminals = getGgTerminalNodeIds(program);
  const foldTargets = getFoldTargets(program);
  const corridorTargets = getCorridorTargets(program);
  const effects = summarizeCapabilityRequirements(
    inferCapabilitiesFromGgSource(normalizedSource)
  );

  const forkEdgeCount = program.edges.filter(
    (edge) => edge.type.toUpperCase() === 'FORK'
  ).length;
  const raceEdgeCount = program.edges.filter(
    (edge) => edge.type.toUpperCase() === 'RACE'
  ).length;

  return {
    moduleName,
    tla: renderTla(
      moduleName,
      program,
      roots,
      terminals,
      foldTargets,
      corridorTargets,
      effects
    ),
    cfg: renderCfg(),
    stats: {
      nodeCount: program.nodes.length,
      edgeCount: program.edges.length,
      rootCount: roots.length,
      terminalCount: terminals.length,
      foldTargetCount: foldTargets.length,
      forkEdgeCount,
      raceEdgeCount,
    },
    effects,
  };
}
