// @a0n/gnosis/src/topological-contagion.ts
/**
 * Topological Contagion & Syzygy
 *
 * Implements algorithms for viral state dissemination across the 
 * dimensional bounds of a Gnostic Mesh Network. Syzygy enables 
 * deep-pairing resonance between connected peers.
 */

import { type MeshConfig } from './spiderweb-topology';

export interface ContagionVector {
  /** The specific localized state being passed */
  payloadId: string;
  /** The topological momentum (how many hops it can still traverse) */
  momentum: number;
  /** Resistance penalty applied at each relay node */
  relayFriction: number;
}

/**
 * Calculates the total potential spread (blast radius) of a topological contagion
 * inside a specific mesh configuration.
 * 
 * In a Syzygy pairing, propagation resistance drops to zero due to perfect resonance.
 */
export function calculateContagionRadius(
  mesh: MeshConfig,
  vector: ContagionVector,
  isSyzygyPairing: boolean
): number {
  if (isSyzygyPairing) {
    // Perfect resonance eliminates relay friction; the contagion spreads
    // unhindered until its raw momentum runs out.
    const maxHops = Math.min(mesh.hops, vector.momentum);
    // V = k * s + k + 1 (Nodes reached)
    return mesh.peers * maxHops + mesh.peers + 1;
  }

  let nodesReached = 1; // Start with the infecting hub
  let currentMomentum = vector.momentum;
  let currentRing = 1;

  // Propagate outward through spiral rings until momentum is lost
  while (currentRing <= mesh.hops && currentMomentum > 0) {
    nodesReached += mesh.peers; // Infect all radials at this ring depth
    
    // Friction slows the contagion
    currentMomentum -= vector.relayFriction;
    currentMomentum -= 1; // baseline hop cost
    currentRing++;
  }

  return nodesReached;
}

/**
 * Checks if a Syzygy alignment is structurally possible between two peers.
 * Requires equal degrees of freedom and complementary spin vectors.
 */
export function checkSyzygyAlignment(
  peerAMomentum: number,
  peerBMomentum: number,
  spinAlignmentDelta: number
): boolean {
  // Resonance requires near-perfect momentum parity and low spin offset
  return Math.abs(peerAMomentum - peerBMomentum) < 0.1 && spinAlignmentDelta < 0.05;
}
