// @a0n/gnosis/src/topological-xor-certificate.ts
/**
 * Topological XOR Certificate
 *
 * Folds finite boundary observations over F2. Even multiplicity cancels,
 * odd multiplicity survives as an unresolved boundary class.
 */

import type { ActivationTrace, Firing } from './scheduling-mesh';

export const TOPOLOGICAL_XOR_CERTIFICATE_VERSION =
  'topological-xor-certificate/v1' as const;

export type TopologicalXorCertificateVersion =
  typeof TOPOLOGICAL_XOR_CERTIFICATE_VERSION;

export type TopologicalXorBoundarySource =
  | 'runtime-trace'
  | 'model-boundary'
  | 'agent-task-graph'
  | 'void-boundary'
  | 'manual';

export type TopologicalXorObservationStatus = 'measured' | 'projected';

export type TopologicalXorCertificateStatus =
  | 'measured'
  | 'projected'
  | 'failed';

export type TopologicalXorProjectionReviewStatus =
  | 'promoted-after-measurement'
  | 'demoted-after-measurement'
  | 'under-investigation';

export type TopologicalXorSkipAction =
  | 'skip-reconstruction'
  | 'skip-with-projection-debt'
  | 'measure-before-skip'
  | 'reconstruct-boundary';

export interface TopologicalXorBoundary {
  /** Stable local identifier for the observed boundary generator. */
  readonly label: string;
  /** Optional coordinate; when present it scopes the parity key. */
  readonly dimension?: number;
  /** Multiplicity in the chain. Defaults to 1. Only parity is retained. */
  readonly magnitude?: number;
  /** Evidence class. Projected inputs keep a balanced certificate projected. */
  readonly status?: TopologicalXorObservationStatus;
  /** Optional origin surface for diagnostics and routing. */
  readonly source?: TopologicalXorBoundarySource;
}

export interface TopologicalXorCertificateInput {
  readonly certificateId: string;
  readonly target: string;
  readonly boundaries: readonly TopologicalXorBoundary[];
  readonly theoremRefs?: readonly string[];
}

export interface TopologicalXorActivationTraceOptions {
  readonly targetPrefix?: string;
  readonly status?: TopologicalXorObservationStatus;
  readonly includePropagateBoundaries?: boolean;
}

export interface TopologicalXorUnresolvedBoundary {
  readonly key: string;
  readonly label: string;
  readonly dimension?: number;
  readonly source?: TopologicalXorBoundarySource;
  readonly parity: 1;
  readonly observations: number;
}

export interface TopologicalXorCertificate {
  readonly version: TopologicalXorCertificateVersion;
  readonly certificateId: string;
  readonly target: string;
  readonly balanced: boolean;
  readonly parityClass: string;
  readonly defectRank: number;
  readonly unresolvedBoundaries: readonly TopologicalXorUnresolvedBoundary[];
  readonly status: TopologicalXorCertificateStatus;
  readonly boundaryCount: number;
  readonly projectedBoundaryCount: number;
  readonly theoremRefs: readonly string[];
}

export interface TopologicalXorProjectionReviewInput {
  readonly reviewId: string;
  readonly projectedCertificate: TopologicalXorCertificate;
  readonly measuredCertificate: TopologicalXorCertificate;
  readonly projectedFidelityPerthou?: number;
  readonly measuredFidelityPerthou?: number;
  readonly passThresholdPerthou?: number;
  readonly theoremRefs?: readonly string[];
}

export interface TopologicalXorProjectionReview {
  readonly reviewId: string;
  readonly projectedCertificateId: string;
  readonly measuredCertificateId: string;
  readonly target: string;
  readonly status: TopologicalXorProjectionReviewStatus;
  readonly passThresholdPerthou: number;
  readonly projectedFidelityPerthou?: number;
  readonly measuredFidelityPerthou?: number;
  readonly projectionErrorPerthou?: number;
  readonly reason: string;
  readonly theoremRefs: readonly string[];
}

export interface TopologicalXorSkipDecisionInput {
  readonly certificate: TopologicalXorCertificate;
  readonly allowProjectedSkip?: boolean;
  readonly requireMeasuredFidelityPerthou?: number;
  readonly measuredFidelityPerthou?: number;
  readonly theoremRefs?: readonly string[];
}

export interface TopologicalXorSkipDecision {
  readonly certificateId: string;
  readonly target: string;
  readonly action: TopologicalXorSkipAction;
  readonly canSkip: boolean;
  readonly speculative: boolean;
  readonly projectionDebt: boolean;
  readonly reason: string;
  readonly defectRank: number;
  readonly parityClass: string;
  readonly unresolvedBoundaryKeys: readonly string[];
  readonly theoremRefs: readonly string[];
}

export interface TopologicalXorSkipPlanInput {
  readonly decisions: readonly TopologicalXorSkipDecision[];
  readonly retentionBudget?: number;
  readonly priorityByTarget?: Readonly<Record<string, number>>;
}

export interface TopologicalXorSkipPlan {
  readonly skipTargets: readonly string[];
  readonly retainedSkipTargets: readonly string[];
  readonly recomputableSkipTargets: readonly string[];
  readonly measureTargets: readonly string[];
  readonly reconstructTargets: readonly string[];
  readonly acceptedProjectionDebtTargets: readonly string[];
  readonly unresolvedBoundaryKeys: readonly string[];
  readonly canSkipAll: boolean;
  readonly hasProjectionDebt: boolean;
  readonly theoremRefs: readonly string[];
}

export interface GrassmannianXorRoundInput {
  readonly roundId: string;
  readonly chart: string;
  readonly laneCount: number;
  readonly activeLaneIndices: readonly number[];
  readonly status?: TopologicalXorObservationStatus;
  readonly theoremRefs?: readonly string[];
}

export const MONSTER_GRASSMANNIAN_LANE_COUNT = 196_884;

export interface PackedGrassmannianLaneMask {
  readonly laneCount: number;
  readonly words: Uint32Array;
}

export const TOPOLOGICAL_XOR_THEOREM_REFS = [
  'Gnosis.InftyOperads.K4_boundary_squared_zero',
  'Gnosis.InftyOperads.K4_H1_zero',
  'Gnosis.OneCobFrobenius.frobenius_identity',
  'Gnosis.BraidCatalogFinal.braid_catalog_final_witness',
] as const;

export const TOPOLOGICAL_XOR_DEMOTION_THEOREM_REFS = [
  'Gnosis.CertificationDemotion.projection_within_family_can_fail_at_scale',
  'Gnosis.CertificationDemotion.qwen_coder_7b_upgrade_conditions_pass',
  'Gnosis.CertificationDemotion.qwen_coder_7b_upgrade_conditions_fail',
  'Gnosis.CertificationDemotion.qwen_coder_7b_upgrade_is_dichotomy',
] as const;

export const TOPOLOGICAL_XOR_SKIP_THEOREM_REFS = [
  'Gnosis.FrfInherentGestalt.xorMerge_append',
  'Gnosis.FrfWitnessTowerTruncation.xor_merge_take_drop_split',
  'Gnosis.FrfWitnessTowerTruncation.xor_parity_track_take_drop_split',
] as const;

interface ParityAccumulator {
  readonly key: string;
  readonly label: string;
  readonly dimension?: number;
  readonly source?: TopologicalXorBoundarySource;
  parity: 0 | 1;
  observations: number;
}

/** Reject empty identifiers before they enter certificate keys. */
function assertNonEmptyString(value: string, fieldName: string): void {
  if (value.trim().length === 0) {
    throw new Error(`${fieldName} must be a non-empty string`);
  }
}

/** Reject non-finite or fractional counts before parity reduction. */
function assertSafeInteger(value: number, fieldName: string): void {
  if (!Number.isSafeInteger(value)) {
    throw new Error(`${fieldName} must be a safe integer`);
  }
}

function assertLaneCount(laneCount: number): void {
  assertSafeInteger(laneCount, 'laneCount');
  if (laneCount <= 0) {
    throw new Error('laneCount must be positive');
  }
}

function assertLaneIndex(laneIndex: number, laneCount: number, fieldName: string): void {
  assertSafeInteger(laneIndex, fieldName);
  if (laneIndex < 0 || laneIndex >= laneCount) {
    throw new Error('active lane index must be inside the lane count');
  }
}

function expectedGrassmannianWordCount(laneCount: number): number {
  return Math.ceil(laneCount / 32);
}

function validatePackedGrassmannianLaneMask(
  mask: PackedGrassmannianLaneMask
): void {
  assertLaneCount(mask.laneCount);
  if (mask.words.length !== expectedGrassmannianWordCount(mask.laneCount)) {
    throw new Error('packed Grassmannian mask word count does not match lane count');
  }
  const unusedBits = mask.words.length * 32 - mask.laneCount;
  if (unusedBits > 0) {
    const validMask = 0xffffffff >>> unusedBits;
    const lastWord = mask.words[mask.words.length - 1] ?? 0;
    if ((lastWord & ~validMask) !== 0) {
      throw new Error('packed Grassmannian mask has bits outside lane count');
    }
  }
}

/** Build the stable F2 generator key for a boundary observation. */
export function topologicalXorBoundaryKey(
  boundary: TopologicalXorBoundary
): string {
  assertNonEmptyString(boundary.label, 'boundary.label');
  if (boundary.dimension === undefined) {
    return boundary.label;
  }
  assertSafeInteger(boundary.dimension, 'boundary.dimension');
  if (boundary.dimension < 0) {
    throw new Error('boundary.dimension must be non-negative');
  }
  return `${boundary.dimension}:${boundary.label}`;
}

/** Reduce a boundary multiplicity to its F2 parity bit. */
export function boundaryMagnitudeParity(
  boundary: TopologicalXorBoundary
): 0 | 1 {
  const magnitude = boundary.magnitude ?? 1;
  assertSafeInteger(magnitude, 'boundary.magnitude');
  return Math.abs(magnitude) % 2 === 0 ? 0 : 1;
}

/** Render the unresolved parity class in deterministic key order. */
function parityClassFromKeys(keys: readonly string[]): string {
  return keys.length === 0 ? '0' : keys.join(' xor ');
}

/** Fold boundary observations into a measured, projected, or failed certificate. */
export function issueTopologicalXorCertificate(
  input: TopologicalXorCertificateInput
): TopologicalXorCertificate {
  assertNonEmptyString(input.certificateId, 'certificateId');
  assertNonEmptyString(input.target, 'target');

  const parityByKey = new Map<string, ParityAccumulator>();
  let projectedBoundaryCount = 0;

  for (const boundary of input.boundaries) {
    const key = topologicalXorBoundaryKey(boundary);
    const parity = boundaryMagnitudeParity(boundary);
    if (boundary.status === 'projected') {
      projectedBoundaryCount += 1;
    }

    const existing = parityByKey.get(key);
    if (existing === undefined) {
      parityByKey.set(key, {
        key,
        label: boundary.label,
        parity,
        observations: 1,
        ...(boundary.dimension !== undefined
          ? { dimension: boundary.dimension }
          : {}),
        ...(boundary.source !== undefined ? { source: boundary.source } : {}),
      });
    } else {
      existing.parity = (existing.parity ^ parity) as 0 | 1;
      existing.observations += 1;
    }
  }

  const unresolvedBoundaries = [...parityByKey.values()]
    .filter((entry) => entry.parity === 1)
    .sort((a, b) => a.key.localeCompare(b.key))
    .map(
      (entry): TopologicalXorUnresolvedBoundary => ({
        key: entry.key,
        label: entry.label,
        parity: 1,
        observations: entry.observations,
        ...(entry.dimension !== undefined ? { dimension: entry.dimension } : {}),
        ...(entry.source !== undefined ? { source: entry.source } : {}),
      })
    );

  const balanced = unresolvedBoundaries.length === 0;
  const status: TopologicalXorCertificateStatus = !balanced
    ? 'failed'
    : projectedBoundaryCount > 0
      ? 'projected'
      : 'measured';

  return {
    version: TOPOLOGICAL_XOR_CERTIFICATE_VERSION,
    certificateId: input.certificateId,
    target: input.target,
    balanced,
    parityClass: parityClassFromKeys(
      unresolvedBoundaries.map((boundary) => boundary.key)
    ),
    defectRank: unresolvedBoundaries.length,
    unresolvedBoundaries,
    status,
    boundaryCount: input.boundaries.length,
    projectedBoundaryCount,
    theoremRefs: [...TOPOLOGICAL_XOR_THEOREM_REFS, ...(input.theoremRefs ?? [])],
  };
}

/** Check internal consistency of a Topological XOR certificate. */
export function verifyTopologicalXorCertificate(
  certificate: TopologicalXorCertificate
): boolean {
  return (
    certificate.version === TOPOLOGICAL_XOR_CERTIFICATE_VERSION &&
    certificate.defectRank === certificate.unresolvedBoundaries.length &&
    certificate.balanced === (certificate.defectRank === 0) &&
    certificate.parityClass ===
      parityClassFromKeys(
        certificate.unresolvedBoundaries.map((boundary) => boundary.key)
      ) &&
    (certificate.status === 'failed') === !certificate.balanced
  );
}

/** Convert per-dimension counts into boundary observations for parity folding. */
export function topologicalXorBoundariesFromCounts(
  counts: readonly number[],
  options: {
    readonly targetPrefix?: string;
    readonly source?: TopologicalXorBoundarySource;
    readonly status?: TopologicalXorObservationStatus;
  } = {}
): TopologicalXorBoundary[] {
  const targetPrefix = options.targetPrefix ?? 'boundary';
  return counts.map((count, dimension) => {
    assertSafeInteger(count, `counts[${dimension}]`);
    return {
      label: `${targetPrefix}-${dimension}`,
      dimension,
      magnitude: count,
      source: options.source ?? 'void-boundary',
      status: options.status ?? 'measured',
    };
  });
}

/**
 * Creates the Packed Grassmannian Lane Mask.
 */
export function createPackedGrassmannianLaneMask(
  laneCount = MONSTER_GRASSMANNIAN_LANE_COUNT
): PackedGrassmannianLaneMask {
  assertLaneCount(laneCount);
  return {
    laneCount,
    words: new Uint32Array(expectedGrassmannianWordCount(laneCount)),
  };
}

/**
 * Handles the gnosis clone Packed Grassmannian Lane Mask workflow.
 */
export function clonePackedGrassmannianLaneMask(
  mask: PackedGrassmannianLaneMask
): PackedGrassmannianLaneMask {
  validatePackedGrassmannianLaneMask(mask);
  return {
    laneCount: mask.laneCount,
    words: new Uint32Array(mask.words),
  };
}

/**
 * Converts input into toggle Packed Grassmannian Lane.
 */
export function togglePackedGrassmannianLane(
  mask: PackedGrassmannianLaneMask,
  laneIndex: number
): void {
  validatePackedGrassmannianLaneMask(mask);
  assertLaneIndex(laneIndex, mask.laneCount, 'laneIndex');
  const wordIndex = laneIndex >>> 5;
  const bit = laneIndex & 31;
  mask.words[wordIndex]! ^= 1 << bit;
}

/**
 * Handles the gnosis xor Packed Grassmannian Lane Masks workflow.
 */
export function xorPackedGrassmannianLaneMasks(
  left: PackedGrassmannianLaneMask,
  right: PackedGrassmannianLaneMask
): PackedGrassmannianLaneMask {
  validatePackedGrassmannianLaneMask(left);
  validatePackedGrassmannianLaneMask(right);
  if (left.laneCount !== right.laneCount) {
    throw new Error('packed Grassmannian masks must use the same lane count');
  }
  const words = new Uint32Array(left.words.length);
  for (let i = 0; i < words.length; i++) {
    words[i] = left.words[i]! ^ right.words[i]!;
  }
  return { laneCount: left.laneCount, words };
}

/**
 * Handles the gnosis packed Grassmannian Lane Mask From Active Lanes workflow.
 */
export function packedGrassmannianLaneMaskFromActiveLanes(
  laneCount: number,
  activeLaneIndices: readonly number[]
): PackedGrassmannianLaneMask {
  const mask = createPackedGrassmannianLaneMask(laneCount);
  for (const laneIndex of activeLaneIndices) {
    togglePackedGrassmannianLane(mask, laneIndex);
  }
  return mask;
}

/**
 * Handles the gnosis packed Grassmannian Lane Mask Defect Rank workflow.
 */
export function packedGrassmannianLaneMaskDefectRank(
  mask: PackedGrassmannianLaneMask
): number {
  validatePackedGrassmannianLaneMask(mask);
  let count = 0;
  for (const word of mask.words) {
    let v = word;
    while (v !== 0) {
      v &= v - 1;
      count++;
    }
  }
  return count;
}

/**
 * Handles the gnosis active Lanes From Packed Grassmannian Lane Mask workflow.
 */
export function activeLanesFromPackedGrassmannianLaneMask(
  mask: PackedGrassmannianLaneMask
): number[] {
  validatePackedGrassmannianLaneMask(mask);
  const lanes: number[] = [];
  for (let wordIndex = 0; wordIndex < mask.words.length; wordIndex++) {
    let word = mask.words[wordIndex]!;
    while (word !== 0) {
      const bit = 31 - Math.clz32(word & -word);
      const laneIndex = wordIndex * 32 + bit;
      if (laneIndex < mask.laneCount) lanes.push(laneIndex);
      word &= word - 1;
    }
  }
  return lanes;
}

/**
 * Issue one F2 parity certificate for a resident cannon round interpreted as a
 * finite Grassmannian chart. Each lane is a Plücker-like coordinate; paired
 * observations cancel, and odd coordinates survive as the changed subspace.
 */
export function issueGrassmannianXorRoundCertificate(
  input: GrassmannianXorRoundInput
): TopologicalXorCertificate {
  assertNonEmptyString(input.roundId, 'roundId');
  assertNonEmptyString(input.chart, 'chart');
  assertLaneCount(input.laneCount);

  const boundaries = input.activeLaneIndices.map(
    (laneIndex): TopologicalXorBoundary => {
      assertLaneIndex(laneIndex, input.laneCount, 'activeLaneIndices[]');
      return {
        label: `${input.chart}:plucker-lane-${laneIndex}`,
        dimension: laneIndex,
        status: input.status ?? 'measured',
        source: 'runtime-trace',
      };
    }
  );

  return issueTopologicalXorCertificate({
    certificateId: `grassmannian-xor-round:${input.roundId}`,
    target: `grassmannian://${input.chart}/round/${input.roundId}`,
    boundaries,
    theoremRefs: [
      'Gnosis.GrassmannianCompiler.amplituhedron_volume_equivalence',
      ...(input.theoremRefs ?? []),
    ],
  });
}

/**
 * Returns whether issue Grassmannian Xor Round Certificate From Mask is true.
 */
export function issueGrassmannianXorRoundCertificateFromMask(input: {
  readonly roundId: string;
  readonly chart: string;
  readonly mask: PackedGrassmannianLaneMask;
  readonly status?: TopologicalXorObservationStatus;
  readonly theoremRefs?: readonly string[];
}): TopologicalXorCertificate {
  validatePackedGrassmannianLaneMask(input.mask);
  return issueGrassmannianXorRoundCertificate({
    roundId: input.roundId,
    chart: input.chart,
    laneCount: input.mask.laneCount,
    activeLaneIndices: activeLanesFromPackedGrassmannianLaneMask(input.mask),
    ...(input.status !== undefined ? { status: input.status } : {}),
    ...(input.theoremRefs !== undefined ? { theoremRefs: input.theoremRefs } : {}),
  });
}

/** Compute absolute projection error in parts per thousand. */
export function projectionErrorPerthou(
  projectedFidelityPerthou: number,
  measuredFidelityPerthou: number
): number {
  assertSafeInteger(projectedFidelityPerthou, 'projectedFidelityPerthou');
  assertSafeInteger(measuredFidelityPerthou, 'measuredFidelityPerthou');
  if (projectedFidelityPerthou < 0 || measuredFidelityPerthou < 0) {
    throw new Error('fidelity perthou values must be non-negative');
  }
  return Math.abs(projectedFidelityPerthou - measuredFidelityPerthou);
}

/** Convert a numerator/denominator fidelity measurement to perthou. */
export function fidelityToPerthou(numerator: number, denominator: number): number {
  assertSafeInteger(numerator, 'numerator');
  assertSafeInteger(denominator, 'denominator');
  if (numerator < 0 || denominator < 0) {
    throw new Error('fidelity inputs must be non-negative');
  }
  return denominator === 0 ? 0 : Math.floor((numerator * 1000) / denominator);
}

/** Promote or demote a projected certificate after measured evidence lands. */
export function reviewProjectedTopologicalXorCertificate(
  input: TopologicalXorProjectionReviewInput
): TopologicalXorProjectionReview {
  assertNonEmptyString(input.reviewId, 'reviewId');
  if (input.projectedCertificate.status !== 'projected') {
    throw new Error('projectedCertificate must have projected status');
  }
  if (input.projectedCertificate.target !== input.measuredCertificate.target) {
    throw new Error('projected and measured certificates must share a target');
  }

  const passThresholdPerthou = input.passThresholdPerthou ?? 950;
  assertSafeInteger(passThresholdPerthou, 'passThresholdPerthou');
  if (passThresholdPerthou < 0) {
    throw new Error('passThresholdPerthou must be non-negative');
  }

  const projectionError =
    input.projectedFidelityPerthou !== undefined &&
    input.measuredFidelityPerthou !== undefined
      ? projectionErrorPerthou(
          input.projectedFidelityPerthou,
          input.measuredFidelityPerthou
        )
      : undefined;

  let status: TopologicalXorProjectionReviewStatus;
  let reason: string;
  if (input.measuredCertificate.status === 'projected') {
    status = 'under-investigation';
    reason = 'measurement certificate still uses projected evidence';
  } else if (input.measuredCertificate.status === 'failed') {
    status = 'demoted-after-measurement';
    reason = 'measured parity certificate failed';
  } else if (
    input.measuredFidelityPerthou !== undefined &&
    input.measuredFidelityPerthou < passThresholdPerthou
  ) {
    status = 'demoted-after-measurement';
    reason = 'measured fidelity missed threshold';
  } else {
    status = 'promoted-after-measurement';
    reason = 'measured parity and fidelity cleared threshold';
  }

  return {
    reviewId: input.reviewId,
    projectedCertificateId: input.projectedCertificate.certificateId,
    measuredCertificateId: input.measuredCertificate.certificateId,
    target: input.projectedCertificate.target,
    status,
    passThresholdPerthou,
    ...(input.projectedFidelityPerthou !== undefined
      ? { projectedFidelityPerthou: input.projectedFidelityPerthou }
      : {}),
    ...(input.measuredFidelityPerthou !== undefined
      ? { measuredFidelityPerthou: input.measuredFidelityPerthou }
      : {}),
    ...(projectionError !== undefined
      ? { projectionErrorPerthou: projectionError }
      : {}),
    reason,
    theoremRefs: [
      ...TOPOLOGICAL_XOR_DEMOTION_THEOREM_REFS,
      ...(input.theoremRefs ?? []),
    ],
  };
}

/**
 * Convert an XOR certificate into a runtime skip decision.
 *
 * A measured balanced certificate may skip reconstruction. A projected balanced
 * certificate may only skip when the caller explicitly accepts projection debt.
 * Failed certificates must reconstruct the unresolved boundary classes.
 */
export function decideTopologicalXorSkip(
  input: TopologicalXorSkipDecisionInput
): TopologicalXorSkipDecision {
  const { certificate } = input;
  if (!verifyTopologicalXorCertificate(certificate)) {
    throw new Error('certificate failed internal consistency check');
  }

  const unresolvedBoundaryKeys = certificate.unresolvedBoundaries.map(
    (boundary) => boundary.key
  );
  const theoremRefs = [
    ...TOPOLOGICAL_XOR_SKIP_THEOREM_REFS,
    ...(input.theoremRefs ?? []),
  ];

  const fidelityRequired = input.requireMeasuredFidelityPerthou;
  if (fidelityRequired !== undefined) {
    assertSafeInteger(fidelityRequired, 'requireMeasuredFidelityPerthou');
    if (fidelityRequired < 0) {
      throw new Error('requireMeasuredFidelityPerthou must be non-negative');
    }
  }
  if (input.measuredFidelityPerthou !== undefined) {
    assertSafeInteger(input.measuredFidelityPerthou, 'measuredFidelityPerthou');
    if (input.measuredFidelityPerthou < 0) {
      throw new Error('measuredFidelityPerthou must be non-negative');
    }
  }

  const base = {
    certificateId: certificate.certificateId,
    target: certificate.target,
    defectRank: certificate.defectRank,
    parityClass: certificate.parityClass,
    unresolvedBoundaryKeys,
    theoremRefs,
  };

  if (!certificate.balanced) {
    return {
      ...base,
      action: 'reconstruct-boundary',
      canSkip: false,
      speculative: false,
      projectionDebt: false,
      reason: `odd boundary classes remain: ${certificate.parityClass}`,
    };
  }

  if (
    fidelityRequired !== undefined &&
    (input.measuredFidelityPerthou === undefined ||
      input.measuredFidelityPerthou < fidelityRequired)
  ) {
    return {
      ...base,
      action: 'measure-before-skip',
      canSkip: false,
      speculative: certificate.status === 'projected',
      projectionDebt: certificate.status === 'projected',
      reason:
        input.measuredFidelityPerthou === undefined
          ? 'measured fidelity is required before skipping reconstruction'
          : 'measured fidelity missed the skip threshold',
    };
  }

  if (certificate.status === 'measured') {
    return {
      ...base,
      action: 'skip-reconstruction',
      canSkip: true,
      speculative: false,
      projectionDebt: false,
      reason: 'measured XOR parity closed all boundary classes',
    };
  }

  if (input.allowProjectedSkip === true) {
    return {
      ...base,
      action: 'skip-with-projection-debt',
      canSkip: true,
      speculative: true,
      projectionDebt: true,
      reason:
        'projected XOR parity closed all boundary classes; caller accepted measurement debt',
    };
  }

  return {
    ...base,
    action: 'measure-before-skip',
    canSkip: false,
    speculative: true,
    projectionDebt: true,
    reason:
      'projected XOR parity is balanced but requires measurement before skipping',
  };
}

/** Partition skip decisions into executable runtime buckets. */
export function planTopologicalXorSkips(
  input: TopologicalXorSkipPlanInput
): TopologicalXorSkipPlan {
  const skipTargets: string[] = [];
  const measureTargets: string[] = [];
  const reconstructTargets: string[] = [];
  const acceptedProjectionDebtTargets: string[] = [];
  const unresolvedBoundaryKeys = new Set<string>();
  const theoremRefs = new Set<string>(TOPOLOGICAL_XOR_SKIP_THEOREM_REFS);
  const retentionBudget = input.retentionBudget ?? input.decisions.length;
  assertSafeInteger(retentionBudget, 'retentionBudget');
  if (retentionBudget < 0) {
    throw new Error('retentionBudget must be non-negative');
  }

  for (const decision of input.decisions) {
    for (const theoremRef of decision.theoremRefs) {
      theoremRefs.add(theoremRef);
    }
    for (const key of decision.unresolvedBoundaryKeys) {
      unresolvedBoundaryKeys.add(key);
    }

    switch (decision.action) {
      case 'skip-reconstruction':
        skipTargets.push(decision.target);
        break;
      case 'skip-with-projection-debt':
        skipTargets.push(decision.target);
        acceptedProjectionDebtTargets.push(decision.target);
        break;
      case 'measure-before-skip':
        measureTargets.push(decision.target);
        break;
      case 'reconstruct-boundary':
        reconstructTargets.push(decision.target);
        break;
    }
  }

  const retainedSkipTargets = skipTargets
    .map((target, index) => ({
      target,
      index,
      priority: input.priorityByTarget?.[target] ?? 0,
    }))
    .sort((left, right) => {
      if (right.priority !== left.priority) {
        return right.priority - left.priority;
      }
      return left.index - right.index;
    })
    .slice(0, retentionBudget)
    .map((entry) => entry.target);
  const retained = new Set(retainedSkipTargets);
  const recomputableSkipTargets = skipTargets.filter(
    (target) => !retained.has(target)
  );

  return {
    skipTargets,
    retainedSkipTargets,
    recomputableSkipTargets,
    measureTargets,
    reconstructTargets,
    acceptedProjectionDebtTargets,
    unresolvedBoundaryKeys: [...unresolvedBoundaryKeys].sort(),
    canSkipAll:
      input.decisions.length > 0 &&
      input.decisions.every((decision) => decision.canSkip),
    hasProjectionDebt: input.decisions.some((decision) => decision.projectionDebt),
    theoremRefs: [...theoremRefs],
  };
}

/** Convert one scheduling-mesh firing into its fork/fold boundary key. */
function activationFiringBoundaryLabel(
  firing: Firing,
  targetPrefix: string
): string | null {
  if (firing.type !== 'fork' && firing.type !== 'fold') {
    return null;
  }
  return `${targetPrefix}:hub:${firing.hub}:hop:${firing.hop}:fork-fold`;
}

/** Convert a scheduling activation trace into F2 boundary observations. */
export function topologicalXorBoundariesFromActivationTrace<T>(
  trace: ActivationTrace<T>,
  options: TopologicalXorActivationTraceOptions = {}
): TopologicalXorBoundary[] {
  const targetPrefix = options.targetPrefix ?? `activation:${trace.startHub}`;
  const status = options.status ?? 'measured';
  const boundaries: TopologicalXorBoundary[] = [];

  for (const firing of trace.firings) {
    const label = activationFiringBoundaryLabel(firing, targetPrefix);
    if (label !== null) {
      boundaries.push({
        label,
        dimension: firing.hop,
        source: 'runtime-trace',
        status,
      });
      continue;
    }

    if (
      options.includePropagateBoundaries === true &&
      firing.type === 'propagate'
    ) {
      boundaries.push({
        label: `${targetPrefix}:propagate:${firing.hub}->${firing.target ?? 'unknown'}:hop:${firing.hop}`,
        dimension: firing.hop,
        source: 'runtime-trace',
        status,
      });
    }
  }

  return boundaries;
}

/** Issue an XOR certificate directly from a scheduling activation trace. */
export function issueActivationTraceXorCertificate<T>(input: {
  readonly certificateId: string;
  readonly target: string;
  readonly trace: ActivationTrace<T>;
  readonly options?: TopologicalXorActivationTraceOptions;
  readonly theoremRefs?: readonly string[];
}): TopologicalXorCertificate {
  return issueTopologicalXorCertificate({
    certificateId: input.certificateId,
    target: input.target,
    boundaries: topologicalXorBoundariesFromActivationTrace(
      input.trace,
      input.options
    ),
    ...(input.theoremRefs !== undefined ? { theoremRefs: input.theoremRefs } : {}),
  });
}
