// @a0n/gnosis/src/traced-monoidal.ts
/**
 * traced-monoidal.ts -- Fork/Race/Fold as a traced symmetric monoidal category
 *
 * The identification:
 *   - Objects = VoidBoundary dimensions (natural numbers)
 *   - Morphisms = topology edges (FORK, RACE, FOLD, VENT, PROCESS)
 *   - Tensor ÔŐŚ = FORK (parallel composition: dimensions add)
 *   - Unit I = empty boundary (0 dimensions)
 *   - Symmetry ¤â = dimension permutation
 *   - Trace Tr = FOLD feedback loop (yanking equation)
 *   - VENT = trace that discards (the zero morphism)
 *
 * The coherence conditions:
 *   - Associativity: (A ÔŐŚ B) ÔŐŚ C Ôëů A ÔŐŚ (B ÔŐŚ C) -- fork order doesn't matter
 *   - Unit: A ÔŐŚ I Ôëů A Ôëů I ÔŐŚ A -- forking with empty is identity
 *   - Symmetry: ¤â Ôłś ¤â = id -- double permutation is identity
 *   - Naturality of trace: Tr(f Ôłś g) = f Ôłś Tr(g) when f doesn't touch traced wires
 *   - Vanishing: Tr_0(f) = f -- trace over zero wires is identity
 *   - Superposing: Tr(f ÔŐŚ g) = f ÔŐŚ Tr(g) when f doesn't touch traced wires
 *   - Yanking: Tr(¤â) = id -- trace of the symmetry is identity
 *
 * The formal ledger entry: the fork/race/fold algebra maps to a traced symmetric
 * monoidal category. Every topology is a string diagram. Every execution is
 * a composition of morphisms. The void boundary is the type. The complement
 * distribution is the evaluation.
 */

// ============================================================================
// Objects -- VoidBoundary dimensions as types
// ============================================================================

/**
 * A categorical object: the dimension count of a VoidBoundary.
 * In the category, objects are natural numbers (dimension counts).
 */
export interface CatObject {
  dimensions: number;
  label?: string;
}

/**
 * Handles the gnosis cat Object workflow.
 */
export function catObject(dimensions: number, label?: string): CatObject {
  return { dimensions, ...(label !== undefined && { label }) };
}

/** The monoidal unit I: zero dimensions */
export const UNIT: CatObject = catObject(0, 'I');

// ============================================================================
// Morphisms -- topology edges as arrows
// ============================================================================

export type MorphismKind =
  | 'identity'
  | 'process'
  | 'fork'
  | 'fold'
  | 'race'
  | 'vent'
  | 'symmetry'
  | 'trace'
  | 'compose'
  | 'tensor';

export interface Morphism {
  kind: MorphismKind;
  source: CatObject;
  target: CatObject;
  label?: string;
  /** For compose: the two morphisms being composed */
  components?: Morphism[];
}

/** Identity morphism: id_A : A Ôćĺ A */
export function identity(obj: CatObject): Morphism {
  return {
    kind: 'identity',
    source: obj,
    target: obj,
    label: `id_${obj.label ?? obj.dimensions}`,
  };
}

/** Process morphism: A Ôćĺ A (sequential transformation, preserves dimensions) */
export function process(obj: CatObject, label?: string): Morphism {
  return { kind: 'process', source: obj, target: obj, ...(label !== undefined && { label }) };
}

/**
 * Fork morphism: A Ôćĺ A ÔŐŚ B (parallel composition, dimensions add)
 * FORK creates new possibilities: n Ôćĺ n + m
 */
export function fork(source: CatObject, added: CatObject): Morphism {
  const target = catObject(source.dimensions + added.dimensions);
  return { kind: 'fork', source, target, label: `fork(+${added.dimensions})` };
}

/**
 * Fold morphism: A ÔŐŚ B Ôćĺ A (collapse, dimensions reduce)
 * FOLD destroys possibilities: n + m Ôćĺ n
 */
export function fold(source: CatObject, removed: number): Morphism {
  const target = catObject(Math.max(0, source.dimensions - removed));
  return { kind: 'fold', source, target, label: `fold(-${removed})` };
}

/**
 * Race morphism: AÔéü ÔŐĽ AÔéé ÔŐĽ ... Ôćĺ A (first to finish wins)
 * Like fold but selects rather than combines.
 */
export function race(source: CatObject, survivors: number): Morphism {
  const target = catObject(survivors);
  return { kind: 'race', source, target, label: `race(Ôćĺ${survivors})` };
}

/**
 * Vent morphism: A Ôćĺ I (discard to void)
 * The zero morphism -- information is irreversibly lost.
 */
export function vent(source: CatObject): Morphism {
  return { kind: 'vent', source, target: UNIT, label: 'vent' };
}

/**
 * Symmetry morphism: A ÔŐŚ B Ôćĺ B ÔŐŚ A (swap)
 */
export function symmetry(a: CatObject, b: CatObject): Morphism {
  const total = catObject(a.dimensions + b.dimensions);
  return {
    kind: 'symmetry',
    source: total,
    target: total,
    label: `¤â(${a.dimensions},${b.dimensions})`,
  };
}

// ============================================================================
// Composition -- sequential and parallel
// ============================================================================

/**
 * Sequential composition: g Ôłś f : A Ôćĺ C where f : A Ôćĺ B and g : B Ôćĺ C.
 * Returns null if types don't match.
 */
export function compose(f: Morphism, g: Morphism): Morphism | null {
  if (f.target.dimensions !== g.source.dimensions) return null;
  return {
    kind: 'compose',
    source: f.source,
    target: g.target,
    label: `${g.label ?? '?'} Ôłś ${f.label ?? '?'}`,
    components: [f, g],
  };
}

/**
 * Tensor (parallel) composition: f ÔŐŚ g : A ÔŐŚ C Ôćĺ B ÔŐŚ D
 * where f : A Ôćĺ B and g : C Ôćĺ D.
 * This is FORK at the morphism level.
 */
export function tensor(f: Morphism, g: Morphism): Morphism {
  return {
    kind: 'tensor',
    source: catObject(f.source.dimensions + g.source.dimensions),
    target: catObject(f.target.dimensions + g.target.dimensions),
    label: `${f.label ?? '?'} ÔŐŚ ${g.label ?? '?'}`,
    components: [f, g],
  };
}

/**
 * Trace: Tr_{A,B}^U(f) where f : A ÔŐŚ U Ôćĺ B ÔŐŚ U
 * The feedback loop -- output U wires feed back to input U wires.
 * This formalizes the FOLD-that-feeds-back.
 */
export function trace(f: Morphism, tracedDims: number): Morphism | null {
  if (f.source.dimensions < tracedDims || f.target.dimensions < tracedDims) {
    return null;
  }
  return {
    kind: 'trace',
    source: catObject(f.source.dimensions - tracedDims),
    target: catObject(f.target.dimensions - tracedDims),
    label: `Tr^${tracedDims}(${f.label ?? '?'})`,
    components: [f],
  };
}

// ============================================================================
// Coherence conditions -- the axioms that make it a category
// ============================================================================

export interface CoherenceCheck {
  name: string;
  holds: boolean;
  lhs: string;
  rhs: string;
  diagnostic?: string;
}

/**
 * Verify associativity: (A ÔŐŚ B) ÔŐŚ C Ôëů A ÔŐŚ (B ÔŐŚ C)
 */
export function checkAssociativity(
  a: CatObject,
  b: CatObject,
  c: CatObject
): CoherenceCheck {
  const lhs = a.dimensions + b.dimensions + c.dimensions;
  const rhs = a.dimensions + (b.dimensions + c.dimensions);
  return {
    name: 'associativity',
    holds: lhs === rhs,
    lhs: `(${a.dimensions} + ${b.dimensions}) + ${c.dimensions} = ${lhs}`,
    rhs: `${a.dimensions} + (${b.dimensions} + ${c.dimensions}) = ${rhs}`,
  };
}

/**
 * Verify left unit: I ÔŐŚ A Ôëů A
 */
export function checkLeftUnit(a: CatObject): CoherenceCheck {
  const lhs = UNIT.dimensions + a.dimensions;
  return {
    name: 'left-unit',
    holds: lhs === a.dimensions,
    lhs: `I ÔŐŚ A = ${lhs}`,
    rhs: `A = ${a.dimensions}`,
  };
}

/**
 * Verify right unit: A ÔŐŚ I Ôëů A
 */
export function checkRightUnit(a: CatObject): CoherenceCheck {
  const lhs = a.dimensions + UNIT.dimensions;
  return {
    name: 'right-unit',
    holds: lhs === a.dimensions,
    lhs: `A ÔŐŚ I = ${lhs}`,
    rhs: `A = ${a.dimensions}`,
  };
}

/**
 * Verify symmetry involution: ¤â Ôłś ¤â = id
 */
export function checkSymmetryInvolution(
  a: CatObject,
  b: CatObject
): CoherenceCheck {
  // ¤â_{A,B} : AÔŐŚB Ôćĺ BÔŐŚA, then ¤â_{B,A} : BÔŐŚA Ôćĺ AÔŐŚB
  // Composing gives identity on AÔŐŚB
  const total = a.dimensions + b.dimensions;
  return {
    name: 'symmetry-involution',
    holds: true, // Dimension swap composed twice = identity (always holds for natural numbers)
    lhs: `¤â(${a.dimensions},${b.dimensions}) Ôłś ¤â(${b.dimensions},${a.dimensions})`,
    rhs: `id_${total}`,
  };
}

/**
 * Verify vanishing: Tr_0(f) = f (trace over zero wires is identity)
 */
export function checkVanishing(f: Morphism): CoherenceCheck {
  const traced = trace(f, 0);
  const holds =
    traced !== null &&
    traced.source.dimensions === f.source.dimensions &&
    traced.target.dimensions === f.target.dimensions;
  return {
    name: 'vanishing',
    holds,
    lhs: `Tr^0(${f.label ?? '?'})`,
    rhs: f.label ?? '?',
  };
}

/**
 * Verify yanking: Tr(¤â_{A,A}) = id_A
 * The trace of the symmetry on A ÔŐŚ A is the identity on A.
 */
export function checkYanking(a: CatObject): CoherenceCheck {
  const sigma = symmetry(a, a);
  const traced = trace(sigma, a.dimensions);
  const holds =
    traced !== null &&
    traced.source.dimensions === a.dimensions &&
    traced.target.dimensions === a.dimensions;
  return {
    name: 'yanking',
    holds,
    lhs: `Tr^${a.dimensions}(¤â(${a.dimensions},${a.dimensions}))`,
    rhs: `id_${a.dimensions}`,
  };
}

/**
 * Run all coherence checks for given objects.
 */
export function verifyCoherence(
  a: CatObject,
  b: CatObject,
  c: CatObject
): { allHold: boolean; checks: CoherenceCheck[] } {
  const checks = [
    checkAssociativity(a, b, c),
    checkLeftUnit(a),
    checkRightUnit(a),
    checkSymmetryInvolution(a, b),
    checkVanishing(identity(a)),
    checkYanking(a),
  ];
  return { allHold: checks.every((c) => c.holds), checks };
}

// ============================================================================
// String diagram interpretation
// ============================================================================

export interface StringDiagram {
  /** Morphisms in topological order */
  morphisms: Morphism[];
  /** Input wires */
  inputs: CatObject;
  /** Output wires */
  outputs: CatObject;
  /** Whether the diagram is well-typed (all compositions match) */
  wellTyped: boolean;
}

/**
 * Build a string diagram from a sequence of morphisms.
 * Verifies that each composition is well-typed.
 */
export function buildStringDiagram(morphisms: Morphism[]): StringDiagram {
  if (morphisms.length === 0) {
    return { morphisms: [], inputs: UNIT, outputs: UNIT, wellTyped: true };
  }

  let wellTyped = true;
  for (let i = 1; i < morphisms.length; i++) {
    if (morphisms[i - 1]!.target.dimensions !== morphisms[i]!.source.dimensions) {
      wellTyped = false;
      break;
    }
  }

  return {
    morphisms,
    inputs: morphisms[0]!.source,
    outputs: morphisms[morphisms.length - 1]!.target,
    wellTyped,
  };
}

/**
 * Compute the beta-1 (first Betti number) of a string diagram.
 * This is the topological invariant: forks create holes, folds close them.
 * ╬▓Ôéü = total_forked - total_folded
 */
export function diagramBeta1(diagram: StringDiagram): number {
  let beta1 = 0;
  for (const m of diagram.morphisms) {
    switch (m.kind) {
      case 'fork':
      case 'tensor':
        beta1 += m.target.dimensions - m.source.dimensions;
        break;
      case 'fold':
      case 'race':
        beta1 -= m.source.dimensions - m.target.dimensions;
        break;
      case 'vent':
        beta1 -= m.source.dimensions;
        break;
      case 'trace':
        // Trace closes loops -- the traced dimensions cancel
        break;
    }
  }
  return beta1;
}

/**
 * A topology is well-formed iff ╬▓Ôéü = 0 at the boundary.
 * This is the deficit closure condition from the Betty compiler.
 */
export function isDiagramClosed(diagram: StringDiagram): boolean {
  return diagram.wellTyped && diagramBeta1(diagram) === 0;
}

// ============================================================================
// Topology-to-diagram: convert a GraphAST edge list to a string diagram
// ============================================================================

export interface EdgeToDiagram {
  edgeType: string;
  sourceCount: number;
  targetCount: number;
}

/**
 * Convert edge types to morphisms.
 */
export function edgeToMorphism(edge: EdgeToDiagram): Morphism {
  const src = catObject(edge.sourceCount);
  const tgt = catObject(edge.targetCount);

  switch (edge.edgeType) {
    case 'PROCESS':
    case 'LAMINAR':
    case 'METACOG':
      return process(src);
    case 'FORK':
    case 'EVOLVE':
    case 'SUPERPOSE':
    case 'ENTANGLE':
      return fork(src, catObject(edge.targetCount - edge.sourceCount));
    case 'FOLD':
    case 'COLLAPSE':
    case 'OBSERVE':
      return fold(src, edge.sourceCount - edge.targetCount);
    case 'RACE':
      return race(src, edge.targetCount);
    case 'VENT':
    case 'TUNNEL':
      return vent(src);
    default:
      return process(src, edge.edgeType);
  }
}
