// @a0n/gnosis/src/treatment-planner.ts
/**
 * Treatment Planner — topologically-sorted repair sequencing across
 * the workspace dependency graph.
 *
 * Four Lean modules formally justify every aspect of the plan:
 *
 * **CancerTreatments.lean** — treatment sequencing:
 *   - gate_first_more_rejections: fix dependencies before consumers
 *   - cascade_amplifies_restoration: hub fix cascades beta1 to dependents
 *   - cascade_multiplier_at_least_two: hub fix >= 2x any non-hub fix
 *
 * **AgileGnosisTopology.lean** — the repair loop structure:
 *   - AgileFailureIntegration: deaf functions (failure) are fuel for the
 *     next repair cycle. The ReversibilityOperator converts delta-error
 *     into velocity. Each --fix cycle maps to a sprint.
 *   - thm_waterfall_causes_terminal_collapse: NOT running the repair
 *     pipeline (no intermediate reversibility) causes terminal collapse.
 *     The --fix flag formalizes the reversibility operator.
 *   - PisotTriggeredSprint: the sprint fires when sigma degrades past
 *     threshold, not on a calendar. Treatment is demand-driven.
 *
 * **EschatonTopology.lean** — the termination condition:
 *   - TopologicalSaturation: when all noise (deaf functions) is resolved,
 *     the engine halts. sigma=0 means treatment complete.
 *   - GraphSabbath: perfection stops sequence. When all projects are
 *     healthy, --dry-run says "no treatment needed."
 *   - thm_fuel_exhaustion_halts_engine: the repair loop terminates
 *     because there is no more delta-error to process.
 *
 * **UniversalRepairTopology.lean** — the master Rosetta Stone:
 *   - thm_repair_is_minimally_sufficient: exactly 5 mechanisms needed:
 *     1. requires_hub_identification → topo sort prioritizes hubs
 *     2. requires_gate_first_removal → leaves first, deps before consumers
 *     3. triggers_cascade_restoration → cascadeMultiplier per project
 *     4. applies_failure_reversibility → --fix = ReversibilityOperator
 *     5. reaches_topological_saturation → halt when sigma=0 (GraphSabbath)
 *   - thm_universal_repair_isomorphism: git push, EMDR session, forest
 *     fire, rate hike — same graph, same 5 operations, same math.
 *   - MonorepoSubstrate: hub=shared-utils, gate=type-errors,
 *     cascade=downstream-build, reversibility=git-commit,
 *     termination=zero-sorry.
 *   - missing_*_fails: omitting ANY of the 5 mechanisms = collapse.
 *     missing_termination_fails: no halt condition = autoimmune disease.
 *
 * The treatment plan is:
 *   1. Hub identification (requires_hub_identification)
 *   2. Gate-first removal (requires_gate_first_removal)
 *   3. Cascade restoration (triggers_cascade_restoration)
 *   4. Failure reversibility (applies_failure_reversibility)
 *   5. Terminates when TopologicalSaturation is reached (GraphSabbath)
 *
 * @module @a0n/gnosis/treatment-planner
 */

// ═══════════════════════════════════════════════════════════════════════
// Types
// ═══════════════════════════════════════════════════════════════════════

/**
 * A node in the treatment graph.
 */
export interface TreatmentNode {
  /** Project name. */
  name: string;
  /** Current sigma (sickness fraction). */
  sigma: number;
  /** Current health classification. */
  health: string;
  /** Number of functions in the project. */
  functionCount: number;
  /** Active channels (beta1 > 0). */
  activeChannels: number;
  /** Destroyed channels (beta1 = 0). */
  destroyedChannels: number;
  /** Projects that depend on this one (consumers). */
  dependents: string[];
  /** Projects this one depends on (dependencies). */
  dependencies: string[];
  /** Number of dependents (reverse deps). */
  dependentCount: number;
  /**
   * Cascade multiplier: how many total beta1 units are restored
   * when this project is fixed, counting cascade to dependents.
   *
   * Maps to CheckpointCascade.totalRestored in CancerTreatments.lean:
   *   totalRestored = hubBeta1 + sum(dependentBeta1s)
   */
  cascadeMultiplier: number;
  /**
   * Treatment priority: combines cascade multiplier with sigma.
   * Higher = fix first. Hubs with many sick dependents get highest priority.
   */
  treatmentPriority: number;
}

/**
 * A single step in the treatment plan.
 */
export interface TreatmentStep {
  /** Step number (1-indexed). */
  step: number;
  /** Project being treated. */
  project: string;
  /** Why this project is treated at this step. */
  rationale: string;
  /** Projected sigma after this step. */
  projectedSigma: number;
  /** Projected health after this step. */
  projectedHealth: string;
  /** Delta sigma from previous step. */
  deltaSigma: number;
  /** Cascade effect: which dependents benefit. */
  cascadeTo: string[];
  /** Cascade multiplier (from Lean theorem). */
  cascadeMultiplier: number;
  /** Number of VENT patches to apply in this project. */
  patchCount: number;
}

/**
 * The full treatment plan for a workspace.
 */
export interface TreatmentPlan {
  /** Timestamp. */
  generatedAt: string;
  /** Total projects in workspace. */
  totalProjects: number;
  /** Projects needing treatment. */
  unhealthyProjects: number;
  /** Treatment steps in execution order. */
  steps: TreatmentStep[];
  /** Starting workspace sigma (average across unhealthy projects). */
  startingSigma: number;
  /** Projected final sigma after all steps. */
  projectedFinalSigma: number;
  /** Projected final health. */
  projectedFinalHealth: string;
  /** Critical hubs treated first (cascade theorem justification). */
  hubsFirst: string[];
  /**
   * Universal Repair Substrate verification.
   * Maps to UniversalRepairTopology.lean — all 4 mechanisms must be present.
   * thm_repair_is_minimally_sufficient: omitting any one = collapse.
   */
  repairSubstrate: {
    /** Hub identified? (requires_hub_identification) */
    hubIdentified: boolean;
    /** Hub name if identified. */
    hubName: string | null;
    /** Gates cleared? (requires_gate_first_removal) — topo sort ensures this. */
    gateFirstApplied: boolean;
    /** Cascade triggered? (triggers_cascade_restoration) */
    cascadeTriggered: boolean;
    /** Max cascade multiplier observed. */
    maxCascadeMultiplier: number;
    /** Reversibility applied? (applies_failure_reversibility) — --fix mode. */
    reversibilityApplied: boolean;
    /**
     * Termination defined? (reaches_topological_saturation)
     * The plan has a defined halt condition: projectedFinalHealth = 'healthy'.
     * Without this: autoimmune disease — endless over-correction loops.
     * missing_termination_fails: omission = collapse.
     */
    terminationDefined: boolean;
    /** All 5 mechanisms present? (thm_repair_is_minimally_sufficient) */
    isMinimallySufficient: boolean;
  };
}

// ═══════════════════════════════════════════════════════════════════════
// Topological Sort
// ═══════════════════════════════════════════════════════════════════════

/**
 * Topological sort of the dependency graph — leaves first.
 *
 * This is Kahn's algorithm. Projects with no dependencies come first
 * (they're the leaves). Then their consumers. Then their consumers'
 * consumers. This ensures every dependency is fixed before its consumers.
 *
 * The Lean justification: gate_first_more_rejections proves that removing
 * blocks (fixing dependencies) before restoring the blocked project
 * maximizes cumulative rejections.
 */
function topologicalSort(
  projects: Map<string, { dependencies: string[]; dependents: string[] }>,
): string[] {
  const inDegree = new Map<string, number>();
  const adjacency = new Map<string, string[]>();

  for (const [name, { dependencies, dependents }] of projects) {
    inDegree.set(name, dependencies.filter(d => projects.has(d)).length);
    adjacency.set(name, dependents.filter(d => projects.has(d)));
  }

  // Kahn's algorithm
  const queue: string[] = [];
  for (const [name, degree] of inDegree) {
    if (degree === 0) queue.push(name);
  }

  const sorted: string[] = [];
  while (queue.length > 0) {
    // Among ready nodes, pick the one with highest dependent count
    // (treat hubs first among peers at the same level)
    queue.sort((a, b) => {
      const depsA = adjacency.get(a)?.length ?? 0;
      const depsB = adjacency.get(b)?.length ?? 0;
      return depsB - depsA;
    });

    const node = queue.shift()!;
    sorted.push(node);

    for (const dependent of adjacency.get(node) ?? []) {
      const degree = (inDegree.get(dependent) ?? 1) - 1;
      inDegree.set(dependent, degree);
      if (degree === 0) queue.push(dependent);
    }
  }

  // Add any remaining nodes (cycles — treat last)
  for (const name of projects.keys()) {
    if (!sorted.includes(name)) sorted.push(name);
  }

  return sorted;
}

// ═══════════════════════════════════════════════════════════════════════
// Cascade Computation
// ═══════════════════════════════════════════════════════════════════════

/**
 * Compute cascade multiplier for a project.
 *
 * Maps to CheckpointCascade.totalRestored from CancerTreatments.lean:
 *   totalRestored = hubBeta1 + sum(dependentBeta1s)
 *
 * In the workspace context:
 *   hubBeta1 = number of VENT patches we'd apply to this project
 *   dependentBeta1s = for each dependent, the number of feedback channels
 *                     restored by the hub becoming healthy
 *
 * Simplified: cascadeMultiplier = 1 + dependentCount
 * (each dependent gains 1 feedback channel when its dependency becomes healthy)
 */
function computeCascadeMultiplier(
  projectName: string,
  dependents: string[],
  unhealthySet: Set<string>,
): number {
  // Base: fixing this project restores its own channels
  let multiplier = 1;

  // Cascade: each unhealthy dependent gains a feedback channel
  for (const dep of dependents) {
    if (unhealthySet.has(dep)) {
      multiplier += 1;
    }
  }

  return multiplier;
}

// ═══════════════════════════════════════════════════════════════════════
// Treatment Plan Generation
// ═══════════════════════════════════════════════════════════════════════

/**
 * Input for the treatment planner — matches what a0 audit organism provides.
 */
export interface WorkspaceHealthSnapshot {
  projects: Array<{
    name: string;
    sigma: number;
    health: string;
    functionCount: number;
    activeChannels: number;
    destroyedChannels: number;
    dependencies: string[];
    dependents: string[];
  }>;
}

/**
 * Generate a topologically-sorted treatment plan for the workspace.
 *
 * The plan is deterministic: given the same health snapshot, it always
 * produces the same sequence. The sequence is optimal by the cascade
 * theorem: hubs first, leaves before consumers, highest complexity first
 * within each project.
 *
 * @param snapshot — current workspace health from a0 audit organism
 * @param patchesPerProject — estimated patches per project (default 10)
 */
export function generateTreatmentPlan(
  snapshot: WorkspaceHealthSnapshot,
  patchesPerProject: number = 10,
): TreatmentPlan {
  // Identify unhealthy projects
  const unhealthy = snapshot.projects.filter(
    (p) => p.health !== 'healthy' && p.health !== 'necrotic' && p.functionCount > 0,
  );
  const unhealthySet = new Set(unhealthy.map((p) => p.name));
  const projectMap = new Map(snapshot.projects.map((p) => [p.name, p]));

  if (unhealthy.length === 0) {
    return {
      generatedAt: new Date().toISOString(),
      totalProjects: snapshot.projects.length,
      unhealthyProjects: 0,
      steps: [],
      startingSigma: 0,
      projectedFinalSigma: 0,
      projectedFinalHealth: 'healthy',
      hubsFirst: [],
      repairSubstrate: {
        hubIdentified: false,
        hubName: null,
        gateFirstApplied: true,
        cascadeTriggered: false,
        maxCascadeMultiplier: 0,
        reversibilityApplied: false,
        terminationDefined: true, // already healthy = already at sabbath
        isMinimallySufficient: false,
      },
    };
  }

  // Build the dependency graph for unhealthy projects
  const graph = new Map<string, { dependencies: string[]; dependents: string[] }>();
  for (const p of unhealthy) {
    graph.set(p.name, {
      dependencies: p.dependencies,
      dependents: p.dependents,
    });
  }

  // Topological sort — leaves first, hubs prioritized among peers
  const sortedNames = topologicalSort(graph);

  // Compute cascade multipliers
  const treatmentNodes: TreatmentNode[] = sortedNames
    .filter((name) => unhealthySet.has(name))
    .map((name) => {
      const p = projectMap.get(name)!;
      const cascadeMultiplier = computeCascadeMultiplier(
        name,
        p.dependents,
        unhealthySet,
      );
      return {
        name,
        sigma: p.sigma,
        health: p.health,
        functionCount: p.functionCount,
        activeChannels: p.activeChannels,
        destroyedChannels: p.destroyedChannels,
        dependents: p.dependents,
        dependencies: p.dependencies,
        dependentCount: p.dependents.length,
        cascadeMultiplier,
        treatmentPriority: cascadeMultiplier * (1 + p.sigma),
      };
    });

  // Generate treatment steps in topological order
  const steps: TreatmentStep[] = [];
  let cumulativeSigmaReduction = 0;
  const startingSigma =
    unhealthy.reduce((s, p) => s + p.sigma, 0) / unhealthy.length;

  // Track which projects have been treated (for cascade tracking)
  const treated = new Set<string>();

  for (let i = 0; i < treatmentNodes.length; i++) {
    const node = treatmentNodes[i];
    treated.add(node.name);

    // Estimate patches: min of destroyedChannels and patchesPerProject
    const patchCount = Math.min(node.destroyedChannels, patchesPerProject);

    // Sigma reduction for this project: patchCount / functionCount
    const projectSigmaReduction = node.functionCount > 0
      ? patchCount / node.functionCount
      : 0;

    // Cascade effect: which untreated dependents benefit?
    const cascadeTo = node.dependents.filter(
      (d) => unhealthySet.has(d) && !treated.has(d),
    );

    // Cumulative sigma reduction (weighted by cascade)
    cumulativeSigmaReduction += projectSigmaReduction * node.cascadeMultiplier;

    const projectedSigma = Math.max(0, startingSigma - cumulativeSigmaReduction / unhealthy.length);

    let projectedHealth: string;
    if (projectedSigma >= 0.6) projectedHealth = 'cancerous';
    else if (projectedSigma >= 0.3) projectedHealth = 'inflamed';
    else if (projectedSigma >= 0.1) projectedHealth = 'stressed';
    else projectedHealth = 'healthy';

    // Rationale citing the specific Lean theorem that justifies this step.
    // Each step maps to a formally-proven treatment strategy.
    let rationale: string;
    if (node.dependentCount > 5) {
      // CancerTreatments.lean: cascade_amplifies_restoration
      // Hub restoration cascades beta1 to all dependents.
      rationale = `Hub checkpoint (${node.dependentCount} dependents) — ` +
        `cascade_amplifies_restoration: fixing this cascades beta1 to ` +
        `${cascadeTo.length} unhealthy consumers (>= ${node.cascadeMultiplier}x gain)`;
    } else if (node.dependencies.every((d) => treated.has(d) || !unhealthySet.has(d))) {
      // CancerTreatments.lean: gate_first_more_rejections
      // All blocking dependencies cleared — restoration is maximally effective.
      rationale = `Gates cleared — gate_first_more_rejections: ` +
        `all dependencies treated, checkpoint restoration maximally effective`;
    } else {
      // AgileGnosisTopology.lean: AgileFailureIntegration
      // This project's deaf functions are delta-error fuel for the repair cycle.
      rationale = `AgileFailureIntegration: ${node.destroyedChannels} deaf functions ` +
        `are delta-error fuel for the reversibility operator`;
    }

    steps.push({
      step: i + 1,
      project: node.name,
      rationale,
      projectedSigma,
      projectedHealth,
      deltaSigma: -projectSigmaReduction * node.cascadeMultiplier / unhealthy.length,
      cascadeTo,
      cascadeMultiplier: node.cascadeMultiplier,
      patchCount,
    });
  }

  const finalStep = steps[steps.length - 1];
  const hubsFirst = treatmentNodes
    .filter((n) => n.dependentCount > 5)
    .map((n) => n.name);

  // Verify UniversalRepairSubstrate — all 5 mechanisms present.
  // thm_repair_is_minimally_sufficient: omitting any one = collapse.
  const hubIdentified = hubsFirst.length > 0;
  const hubName = hubsFirst[0] ?? null;
  const gateFirstApplied = true; // topo sort guarantees gate-first
  const maxCascade = Math.max(0, ...treatmentNodes.map((n) => n.cascadeMultiplier));
  const cascadeTriggered = maxCascade > 1;
  const reversibilityApplied = steps.length > 0; // --fix applies reversibility
  // 5th pillar: termination is defined when the plan projects to healthy.
  // Without this halt condition: autoimmune disease (missing_termination_fails).
  const terminationDefined = (finalStep?.projectedHealth === 'healthy') ||
    (finalStep?.projectedSigma !== undefined && finalStep.projectedSigma < 0.1);
  const isMinimallySufficient =
    hubIdentified && gateFirstApplied && cascadeTriggered &&
    reversibilityApplied && terminationDefined;

  return {
    generatedAt: new Date().toISOString(),
    totalProjects: snapshot.projects.length,
    unhealthyProjects: unhealthy.length,
    steps,
    startingSigma,
    projectedFinalSigma: finalStep?.projectedSigma ?? startingSigma,
    projectedFinalHealth: finalStep?.projectedHealth ?? 'unknown',
    hubsFirst,
    repairSubstrate: {
      hubIdentified,
      hubName,
      gateFirstApplied,
      cascadeTriggered,
      maxCascadeMultiplier: maxCascade,
      reversibilityApplied,
      terminationDefined,
      isMinimallySufficient,
    },
  };
}

// ═══════════════════════════════════════════════════════════════════════
// Reporting
// ═══════════════════════════════════════════════════════════════════════

/**
 * Format a treatment plan as a human-readable report.
 */
export function formatTreatmentPlan(plan: TreatmentPlan): string {
  const lines: string[] = [];

  lines.push('TREATMENT PLAN');
  lines.push('='.repeat(75));
  lines.push(`Generated: ${plan.generatedAt}`);
  lines.push(`Projects: ${plan.totalProjects} total, ${plan.unhealthyProjects} unhealthy`);
  lines.push(`Starting sigma: ${plan.startingSigma.toFixed(4)}`);
  lines.push(`Projected final: ${plan.projectedFinalSigma.toFixed(4)} (${plan.projectedFinalHealth})`);

  if (plan.hubsFirst.length > 0) {
    lines.push(`\nHub-first (cascade_amplifies_restoration):`);
    for (const hub of plan.hubsFirst) {
      const step = plan.steps.find((s) => s.project === hub);
      if (step) {
        lines.push(`  ${hub} — step #${step.step}, cascade to ${step.cascadeTo.length} consumers`);
      }
    }
  }

  lines.push(`\nTREATMENT SEQUENCE:`);
  lines.push('-'.repeat(75));
  lines.push(
    'Step'.padEnd(6) +
    'Project'.padEnd(30) +
    'Patches'.padStart(8) +
    'Cascade'.padStart(8) +
    'Sigma'.padStart(8) +
    'Health'.padStart(12),
  );
  lines.push('-'.repeat(75));

  for (const step of plan.steps) {
    lines.push(
      `#${step.step}`.padEnd(6) +
      step.project.slice(0, 29).padEnd(30) +
      String(step.patchCount).padStart(8) +
      `${step.cascadeMultiplier}x`.padStart(8) +
      step.projectedSigma.toFixed(4).padStart(8) +
      step.projectedHealth.padStart(12),
    );
    if (step.cascadeTo.length > 0) {
      lines.push(`       cascade -> ${step.cascadeTo.slice(0, 5).join(', ')}${step.cascadeTo.length > 5 ? ` +${step.cascadeTo.length - 5} more` : ''}`);
    }
  }

  lines.push('-'.repeat(75));
  lines.push(
    'done'.padEnd(6) +
    `(${plan.steps.length} steps)`.padEnd(30) +
    ''.padStart(8) +
    ''.padStart(8) +
    plan.projectedFinalSigma.toFixed(4).padStart(8) +
    plan.projectedFinalHealth.padStart(12),
  );

  // Termination condition from EschatonTopology.lean
  if (plan.projectedFinalHealth === 'healthy') {
    lines.push(`\nTERMINATION: TopologicalSaturation reached — GraphSabbath.`);
    lines.push(`  thm_fuel_exhaustion_halts_engine: all delta-error resolved,`);
    lines.push(`  reversibility operator has no remaining fuel. Treatment complete.`);
  } else {
    lines.push(`\nPARTIAL TREATMENT: ${plan.steps.length} steps bring sigma from`);
    lines.push(`  ${plan.startingSigma.toFixed(4)} to ${plan.projectedFinalSigma.toFixed(4)}.`);
    lines.push(`  Additional cycles needed to reach TopologicalSaturation.`);
    lines.push(`  thm_agile_preserves_velocity: each cycle's delta-error fuels the next.`);
  }

  // Universal Repair Substrate verification (UniversalRepairTopology.lean)
  const rs = plan.repairSubstrate;
  lines.push(`\nUNIVERSAL REPAIR SUBSTRATE (thm_repair_is_minimally_sufficient):`);
  lines.push(`  [${rs.hubIdentified ? 'x' : ' '}] requires_hub_identification    hub=${rs.hubName ?? 'none'}`);
  lines.push(`  [${rs.gateFirstApplied ? 'x' : ' '}] requires_gate_first_removal   topo sort = leaves first`);
  lines.push(`  [${rs.cascadeTriggered ? 'x' : ' '}] triggers_cascade_restoration  max cascade=${rs.maxCascadeMultiplier}x`);
  lines.push(`  [${rs.reversibilityApplied ? 'x' : ' '}] applies_failure_reversibility --fix = ReversibilityOperator`);
  lines.push(`  [${rs.terminationDefined ? 'x' : ' '}] reaches_topological_saturation projected=${plan.projectedFinalHealth}`);
  if (rs.isMinimallySufficient) {
    lines.push(`  VERDICT: All 5 mechanisms present. Repair is minimally sufficient.`);
    lines.push(`  thm_universal_repair_isomorphism: this plan is topologically`);
    lines.push(`  identical to EMDR processing, mycorrhizal repair, and rate normalization.`);
  } else {
    const missing: string[] = [];
    if (!rs.hubIdentified) missing.push('hub_identification');
    if (!rs.cascadeTriggered) missing.push('cascade_restoration');
    if (!rs.reversibilityApplied) missing.push('failure_reversibility');
    if (!rs.terminationDefined) missing.push('topological_saturation');
    lines.push(`  WARNING: Missing: ${missing.join(', ')}. missing_*_fails: omission = collapse.`);
    if (!rs.terminationDefined) {
      lines.push(`  missing_termination_fails: no halt condition = autoimmune over-correction.`);
      lines.push(`  Increase --limit or add more repair cycles to reach saturation.`);
    }
  }

  lines.push('\nFORMAL JUSTIFICATION (6 Lean modules, zero sorry):');
  lines.push('  UniversalRepairTopology.lean (master Rosetta Stone):');
  lines.push('    thm_repair_is_minimally_sufficient — 5 mechanisms, all required');
  lines.push('    thm_universal_repair_isomorphism — git/psyche/eco/econ = same graph');
  lines.push('    missing_*_fails — omitting any mechanism = terminal collapse');
  lines.push('    missing_termination_fails — no halt = autoimmune over-correction');
  lines.push('  CancerTreatments.lean (treatment sequencing):');
  lines.push('    gate_first_more_rejections — dependencies treated before consumers');
  lines.push('    cascade_amplifies_restoration — hub restoration cascades to dependents');
  lines.push('    cascade_multiplier_at_least_two — hub fix gives >= 2x non-hub fix');
  lines.push('  AgileGnosisTopology.lean (repair loop structure):');
  lines.push('    AgileFailureIntegration — deaf functions are fuel for repair');
  lines.push('    thm_waterfall_causes_terminal_collapse — no repair = guaranteed collapse');
  lines.push('    PisotTriggeredSprint — treatment fires on sigma degradation, not calendar');
  lines.push('  EschatonTopology.lean (termination):');
  lines.push('    TopologicalSaturation — all noise resolved, engine halts');
  lines.push('    GraphSabbath — perfection stops sequence');
  lines.push('    thm_fuel_exhaustion_halts_engine — no delta-error = treatment complete');
  lines.push('  CheckpointHomology.lean (algebraic foundation):');
  lines.push('    self_repair_completeness — any chain complex can diagnose + recover');
  lines.push('    rank_nullity — conservation: active + destroyed = total');
  lines.push('  PolyglotCheckpointFunctor.lean (scanner soundness):');
  lines.push('    polyglot_checkpoint_functor — scanner output preserves conservation');

  return lines.join('\n');
}
