// @a0n/gnosis/src/ts-call-graph.ts
/**
 * Gnosis Cross-Function Call Graph
 *
 * Walks the TypeScript import graph and builds a module-level GG topology
 * where each exported function is a node and calls between them are edges.
 * Computes module-level buley, wallace, and regime.
 */

import * as fs from 'node:fs';
import * as path from 'node:path';
import { createRequire } from 'node:module';
import type * as TsTypes from 'typescript';
import {
  compileTypeScriptToGnosis,
//   type _GnosisTypeScriptBridgeResult,
  type GnosisTypeScriptBridgeSourceLocation,
} from './ts-bridge';
import {
  analyzeGnosisSource,
//   _formatGnosisViolations,
  type GnosisComplexityReport,
  type GnosisSteeringRegime,
} from './analysis';

type TypeScriptModule = typeof import('typescript');

function resolveRuntimeRequireBase(): string | null {
  if (
    typeof __filename === 'string' &&
    __filename.length > 0 &&
    path.isAbsolute(__filename)
  ) {
    return __filename;
  }

  const moduleUrl = import.meta.url;
  return typeof moduleUrl === 'string' && moduleUrl.startsWith('file:')
    ? moduleUrl
    : null;
}

const runtimeRequireBase = resolveRuntimeRequireBase();
const runtimeRequire = runtimeRequireBase
  ? createRequire(runtimeRequireBase)
  : null;

let cachedTs: TypeScriptModule | undefined;

function loadTs(): TypeScriptModule {
  if (runtimeRequire === null) {
    throw new Error(
      'TypeScript call-graph analysis requires a Node-compatible module path in this runtime.'
    );
  }

  if (!cachedTs) {
    cachedTs = runtimeRequire('typescript') as TypeScriptModule;
  }
  return cachedTs;
}

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface CallGraphNode {
  readonly id: string;
  readonly filePath: string;
  readonly functionName: string;
  readonly exportName: string;
  readonly sourceLocation: GnosisTypeScriptBridgeSourceLocation;
  readonly buleyNumber: number;
  readonly wallaceNumber: number;
  readonly regime: GnosisSteeringRegime;
  readonly beta1: number;
  readonly nodeCount: number;
  readonly edgeCount: number;
}

export interface CallGraphEdge {
  readonly from: string;
  readonly to: string;
  readonly callCount: number;
  readonly sourceLocations: readonly GnosisTypeScriptBridgeSourceLocation[];
}

export interface CallGraphModule {
  readonly filePath: string;
  readonly functions: readonly CallGraphNode[];
  readonly internalCalls: readonly CallGraphEdge[];
  readonly externalImports: readonly CallGraphExternalImport[];
}

export interface CallGraphExternalImport {
  readonly specifier: string;
  readonly resolvedPath: string | null;
  readonly importedNames: readonly string[];
  readonly sourceLocation: GnosisTypeScriptBridgeSourceLocation;
}

export interface CallGraph {
  readonly modules: readonly CallGraphModule[];
  readonly nodes: readonly CallGraphNode[];
  readonly edges: readonly CallGraphEdge[];
  readonly composedBuley: number;
  readonly composedWallace: number;
  readonly composedRegime: GnosisSteeringRegime;
  readonly composedBeta1: number;
  readonly composedGgSource: string;
}

export interface CallGraphOptions {
  readonly entryFiles: readonly string[];
  readonly maxDepth?: number;
  readonly followImports?: boolean;
}

// ---------------------------------------------------------------------------
// Import resolution
// ---------------------------------------------------------------------------

const TS_EXTENSIONS = ['.ts', '.tsx', '.mts', '.cts'] as const;
const ALL_EXTENSIONS = [
  '.ts',
  '.tsx',
  '.mts',
  '.cts',
  '.js',
  '.jsx',
  '.mjs',
  '.cjs',
] as const;

function resolveImportPath(specifier: string, fromFile: string): string | null {
  if (!specifier.startsWith('./') && !specifier.startsWith('../')) {
    return null; // bare specifier -- skip
  }

  const dir = path.dirname(path.resolve(fromFile));
  const base = path.resolve(dir, specifier);

  // Try with existing extension
  if (fs.existsSync(base) && fs.statSync(base).isFile()) {
    return base;
  }

  // Try adding extensions
  for (const ext of ALL_EXTENSIONS) {
    const candidate = base.endsWith(ext) ? base : `${base}${ext}`;
    if (fs.existsSync(candidate)) {
      return candidate;
    }
  }

  // Try index files
  for (const ext of ALL_EXTENSIONS) {
    const candidate = path.join(base, `index${ext}`);
    if (fs.existsSync(candidate)) {
      return candidate;
    }
  }

  return null;
}

// ---------------------------------------------------------------------------
// TS AST extraction
// ---------------------------------------------------------------------------

interface ExtractedFunction {
  name: string;
  exportName: string;
  line: number;
  column: number;
  endLine: number;
  endColumn: number;
  calledFunctions: readonly string[];
  isAsync: boolean;
}

interface ExtractedImport {
  specifier: string;
  importedNames: readonly string[];
  line: number;
  column: number;
}

interface FileExtraction {
  filePath: string;
  functions: readonly ExtractedFunction[];
  imports: readonly ExtractedImport[];
}

function extractFileInfo(filePath: string, sourceText: string): FileExtraction {
  const ts = loadTs();
  const scriptKind = filePath.endsWith('.tsx')
    ? ts.ScriptKind.TSX
    : filePath.endsWith('.jsx')
    ? ts.ScriptKind.JSX
    : ts.ScriptKind.TS;

  const sourceFile = ts.createSourceFile(
    filePath,
    sourceText,
    ts.ScriptTarget.Latest,
    true,
    scriptKind
  );

  const functions: ExtractedFunction[] = [];
  const imports: ExtractedImport[] = [];
  const topLevelFunctionNames = new Set<string>();

  // First pass: collect all top-level function declarations
  for (const statement of sourceFile.statements) {
    if (ts.isFunctionDeclaration(statement) && statement.name) {
      topLevelFunctionNames.add(statement.name.text);
    }
    if (ts.isVariableStatement(statement)) {
      for (const decl of statement.declarationList.declarations) {
        if (
          ts.isIdentifier(decl.name) &&
          decl.initializer &&
          (ts.isArrowFunction(decl.initializer) ||
            ts.isFunctionExpression(decl.initializer))
        ) {
          topLevelFunctionNames.add(decl.name.text);
        }
      }
    }
  }

  // Second pass: extract function details
  for (const statement of sourceFile.statements) {
    // Import declarations
    if (ts.isImportDeclaration(statement) && statement.moduleSpecifier) {
      const specifier = ts.isStringLiteralLike(statement.moduleSpecifier)
        ? statement.moduleSpecifier.text
        : '';
      const importedNames: string[] = [];

      if (statement.importClause) {
        if (statement.importClause.name) {
          importedNames.push(statement.importClause.name.text);
        }
        if (
          statement.importClause.namedBindings &&
          ts.isNamedImports(statement.importClause.namedBindings)
        ) {
          for (const element of statement.importClause.namedBindings.elements) {
            importedNames.push(element.name.text);
          }
        }
      }

      const pos = sourceFile.getLineAndCharacterOfPosition(
        statement.getStart(sourceFile)
      );
      imports.push({
        specifier,
        importedNames,
        line: pos.line + 1,
        column: pos.character + 1,
      });
    }

    // Exported function declarations
    const isExported =
      ts.canHaveModifiers(statement) &&
      ts
        .getModifiers(statement)
        ?.some((m) => m.kind === ts.SyntaxKind.ExportKeyword);

    if (
      ts.isFunctionDeclaration(statement) &&
      statement.name &&
      statement.body
    ) {
      const name = statement.name.text;
      const calledFunctions = collectCalledFunctions(
        statement.body,
        sourceFile,
        ts,
        topLevelFunctionNames
      );
      const start = sourceFile.getLineAndCharacterOfPosition(
        statement.getStart(sourceFile)
      );
      const end = sourceFile.getLineAndCharacterOfPosition(statement.getEnd());

      functions.push({
        name,
        exportName: isExported ? name : `_${name}`,
        line: start.line + 1,
        column: start.character + 1,
        endLine: end.line + 1,
        endColumn: end.character + 1,
        calledFunctions,
        isAsync:
          ts.canHaveModifiers(statement) &&
          (ts
            .getModifiers(statement)
            ?.some((m) => m.kind === ts.SyntaxKind.AsyncKeyword) ??
            false),
      });
    }

    if (ts.isVariableStatement(statement) && isExported) {
      for (const decl of statement.declarationList.declarations) {
        if (
          ts.isIdentifier(decl.name) &&
          decl.initializer &&
          (ts.isArrowFunction(decl.initializer) ||
            ts.isFunctionExpression(decl.initializer))
        ) {
          const name = decl.name.text;
          const body = ts.isArrowFunction(decl.initializer)
            ? decl.initializer.body
            : decl.initializer.body;
          const calledFunctions = ts.isBlock(body)
            ? collectCalledFunctions(
                body,
                sourceFile,
                ts,
                topLevelFunctionNames
              )
            : [];
          const start = sourceFile.getLineAndCharacterOfPosition(
            decl.getStart(sourceFile)
          );
          const end = sourceFile.getLineAndCharacterOfPosition(decl.getEnd());

          functions.push({
            name,
            exportName: name,
            line: start.line + 1,
            column: start.character + 1,
            endLine: end.line + 1,
            endColumn: end.character + 1,
            calledFunctions,
            isAsync:
              ts.canHaveModifiers(decl.initializer) &&
              (ts
                .getModifiers(decl.initializer)
                ?.some(
                  (m: TsTypes.Modifier) => m.kind === ts.SyntaxKind.AsyncKeyword
                ) ??
                false),
          });
        }
      }
    }
  }

  return { filePath, functions, imports };
}

function collectCalledFunctions(
  node: TsTypes.Node,
  _sourceFile: TsTypes.SourceFile,
  ts: TypeScriptModule,
  knownFunctions: Set<string>
): string[] {
  const calls: string[] = [];

  const visit = (child: TsTypes.Node): void => {
    if (ts.isCallExpression(child) && ts.isIdentifier(child.expression)) {
      const name = child.expression.text;
      if (knownFunctions.has(name)) {
        calls.push(name);
      }
    }
    ts.forEachChild(child, visit);
  };

  ts.forEachChild(node, visit);
  return [...new Set(calls)];
}

// ---------------------------------------------------------------------------
// Call graph construction
// ---------------------------------------------------------------------------

/**
 * Builds the Call Graph.
 */
export async function buildCallGraph(
  options: CallGraphOptions
): Promise<CallGraph> {
  const maxDepth = options.maxDepth ?? 5;
  const followImports = options.followImports ?? true;
  const visited = new Set<string>();
  const allModules: CallGraphModule[] = [];

  async function processFile(filePath: string, depth: number): Promise<void> {
    const absPath = path.resolve(filePath);
    if (visited.has(absPath) || depth > maxDepth) {
      return;
    }
    visited.add(absPath);

    if (!fs.existsSync(absPath)) {
      return;
    }

    // polyglot:ignore RESOURCE_LEAK — readFileSync returns a string, no handle to release
    const sourceText = fs.readFileSync(absPath, 'utf-8');
    const extraction = extractFileInfo(absPath, sourceText);
    const nodes: CallGraphNode[] = [];
    const internalCalls: CallGraphEdge[] = [];
    const externalImports: CallGraphExternalImport[] = [];

    // Build per-function analysis
    for (const func of extraction.functions) {
      let buley = 0;
      let wallace = 0;
      let regime: GnosisSteeringRegime = 'laminar';
      let beta1 = 0;
      let nodeCount = 0;
      let edgeCount = 0;

      try {
        const bridgeResult = compileTypeScriptToGnosis(sourceText, {
          sourceFilePath: absPath,
          ...(func.exportName.startsWith('_')
            ? {}
            : { exportName: func.exportName }),
        });

        const report = await analyzeGnosisSource(bridgeResult.ggSource);
        buley = report.buleyNumber;
        wallace = report.steering.wallaceNumber;
        regime = report.steering.regime;
        beta1 = report.topology.structuralBeta1;
        nodeCount = report.topology.nodeCount;
        edgeCount = report.topology.edgeCount;
      } catch {
        // Functions that can't be bridged get zero metrics
      }

      const nodeId = `${path.basename(absPath, path.extname(absPath))}::${
        func.name
      }`;
      nodes.push({
        id: nodeId,
        filePath: absPath,
        functionName: func.name,
        exportName: func.exportName,
        sourceLocation: {
          line: func.line,
          column: func.column,
          endLine: func.endLine,
          endColumn: func.endColumn,
        },
        buleyNumber: buley,
        wallaceNumber: wallace,
        regime,
        beta1,
        nodeCount,
        edgeCount,
      });

      // Internal call edges
      for (const calledName of func.calledFunctions) {
        const targetNode = extraction.functions.find(
          (f) => f.name === calledName
        );
        if (targetNode) {
          const targetId = `${path.basename(
            absPath,
            path.extname(absPath)
          )}::${calledName}`;
          const existing = internalCalls.find(
            (e) => e.from === nodeId && e.to === targetId
          );
          if (existing) {
            // increment count -- but readonly, so skip
          } else {
            internalCalls.push({
              from: nodeId,
              to: targetId,
              callCount: 1,
              sourceLocations: [
                {
                  line: func.line,
                  column: func.column,
                },
              ],
            });
          }
        }
      }
    }

    // Process imports
    for (const imp of extraction.imports) {
      const resolvedPath = resolveImportPath(imp.specifier, absPath);
      externalImports.push({
        specifier: imp.specifier,
        resolvedPath,
        importedNames: imp.importedNames,
        sourceLocation: {
          line: imp.line,
          column: imp.column,
        },
      });

      // Follow imports
      if (followImports && resolvedPath) {
        const ext = path.extname(resolvedPath);
        if (TS_EXTENSIONS.includes(ext as (typeof TS_EXTENSIONS)[number])) {
          await processFile(resolvedPath, depth + 1);
        }
      }
    }

    allModules.push({
      filePath: absPath,
      functions: nodes,
      internalCalls,
      externalImports,
    });
  }

  for (const file of options.entryFiles) {
    await processFile(file, 0);
  }

  // Build cross-module edges
  const allNodes = allModules.flatMap((m) => m.functions);
  const allEdges: CallGraphEdge[] = [
    ...allModules.flatMap((m) => m.internalCalls),
  ];

  // Resolve cross-module call edges
  for (const module of allModules) {
    for (const imp of module.externalImports) {
      if (!imp.resolvedPath) continue;

      const targetModule = allModules.find(
        (m) => m.filePath === imp.resolvedPath
      );
      if (!targetModule) continue;

      for (const importedName of imp.importedNames) {
        const targetNode = targetModule.functions.find(
          (n) =>
            n.functionName === importedName || n.exportName === importedName
        );
        if (!targetNode) continue;

        // Find callers in this module that use this import
        for (const func of module.functions) {
          const nodeId = func.id;
          const existingEdge = allEdges.find(
            (e) => e.from === nodeId && e.to === targetNode.id
          );
          if (!existingEdge) {
            allEdges.push({
              from: nodeId,
              to: targetNode.id,
              callCount: 1,
              sourceLocations: [imp.sourceLocation],
            });
          }
        }
      }
    }
  }

  // Compose GG source for the full call graph
  const ggLines = ['// call-graph.gg — composed module topology'];
  for (const node of allNodes) {
    ggLines.push(
      `(${sanitizeNodeId(node.id)}:Function { buley: '${
        node.buleyNumber
      }', regime: '${node.regime}' })`
    );
  }
  for (const edge of allEdges) {
    ggLines.push(
      `(${sanitizeNodeId(edge.from)})-[:PROCESS]->(${sanitizeNodeId(edge.to)})`
    );
  }
  const composedGgSource = ggLines.join('\n') + '\n';

  // Compute composed metrics
  let composedReport: GnosisComplexityReport | null = null;
  try {
    composedReport = await analyzeGnosisSource(composedGgSource);
  } catch {
    // Composed topology may not parse if there are no edges
  }

  return {
    modules: allModules,
    nodes: allNodes,
    edges: allEdges,
    composedBuley:
      composedReport?.buleyNumber ?? sumMetric(allNodes, 'buleyNumber'),
    composedWallace:
      composedReport?.steering.wallaceNumber ??
      maxMetric(allNodes, 'wallaceNumber'),
    composedRegime: composedReport?.steering.regime ?? worstRegime(allNodes),
    composedBeta1:
      composedReport?.topology.structuralBeta1 ?? maxMetric(allNodes, 'beta1'),
    composedGgSource,
  };
}

function sanitizeNodeId(id: string): string {
  return id.replace(/[^A-Za-z0-9_]/g, '_');
}

function sumMetric(
  nodes: readonly CallGraphNode[],
  key: 'buleyNumber'
): number {
  return nodes.reduce((sum, n) => sum + n[key], 0);
}

function maxMetric(
  nodes: readonly CallGraphNode[],
  key: 'wallaceNumber' | 'beta1'
): number {
  return nodes.reduce((max, n) => Math.max(max, n[key]), 0);
}

function worstRegime(nodes: readonly CallGraphNode[]): GnosisSteeringRegime {
  if (nodes.some((n) => n.regime === 'turbulent')) return 'turbulent';
  if (nodes.some((n) => n.regime === 'transitional')) return 'transitional';
  return 'laminar';
}
