// @a0n/gnosis/src/ts-runtime-probe.ts
/**
 * Gnosis Runtime Topology Probes
 *
 * Lightweight instrumentation injected at topology node boundaries.
 * Records execution paths, timings per node, and branch selection frequencies.
 * Correlates runtime data with static topology for empirical wallace surfaces.
 */

export interface ProbeEvent {
  readonly nodeId: string;
  readonly kind: 'enter' | 'exit' | 'fork-branch' | 'fold-complete' | 'error';
  readonly timestamp: number;
  readonly durationMs?: number;
  readonly metadata?: Record<string, unknown>;
}

export interface ProbeSession {
  readonly sessionId: string;
  readonly functionName: string;
  readonly filePath: string;
  readonly startedAt: number;
  readonly events: readonly ProbeEvent[];
  readonly completedAt: number | null;
}

export interface RuntimeTopologyMetrics {
  readonly functionName: string;
  readonly filePath: string;
  readonly invocationCount: number;
  readonly avgDurationMs: number;
  readonly p50DurationMs: number;
  readonly p95DurationMs: number;
  readonly p99DurationMs: number;
  readonly nodeHotspots: readonly NodeHotspot[];
  readonly branchFrequencies: readonly BranchFrequency[];
  readonly empiricalWallace: number;
  readonly empiricalRegime: 'laminar' | 'transitional' | 'turbulent';
}

export interface NodeHotspot {
  readonly nodeId: string;
  readonly avgDurationMs: number;
  readonly invocationCount: number;
  readonly errorCount: number;
  readonly percentOfTotal: number;
}

export interface BranchFrequency {
  readonly forkNodeId: string;
  readonly branchNodeId: string;
  readonly selectionCount: number;
  readonly selectionRate: number;
}

export interface WallaceDivergence {
  readonly functionName: string;
  readonly staticWallace: number;
  readonly empiricalWallace: number;
  readonly divergence: number;
  readonly interpretation: 'pessimistic-model' | 'aligned' | 'latent-risk';
}

// ---------------------------------------------------------------------------
// Probe collector
// ---------------------------------------------------------------------------

export class GnosisProbeCollector {
  private sessions = new Map<string, MutableSession>();
  private completedSessions: MutableSession[] = [];
  private readonly maxCompleted: number;
  private sessionCounter = 0;

  constructor(maxCompleted = 10000) {
    this.maxCompleted = maxCompleted;
  }

  startSession(functionName: string, filePath: string): string {
    const sessionId = `probe_${++this.sessionCounter}_${Date.now()}`;
    this.sessions.set(sessionId, {
      sessionId,
      functionName,
      filePath,
      startedAt: Date.now(),
      events: [],
      completedAt: null,
    });
    return sessionId;
  }

  recordEvent(sessionId: string, event: Omit<ProbeEvent, 'timestamp'>): void {
    const session = this.sessions.get(sessionId);
    if (!session) return;

    session.events.push({
      ...event,
      timestamp: Date.now(),
    });
  }

  endSession(sessionId: string): void {
    const session = this.sessions.get(sessionId);
    if (!session) return;

    session.completedAt = Date.now();
    this.sessions.delete(sessionId);
    this.completedSessions.push(session);

    // Evict old sessions
    while (this.completedSessions.length > this.maxCompleted) {
      this.completedSessions.shift();
    }
  }

  getMetrics(functionName: string, filePath: string): RuntimeTopologyMetrics {
    const sessions = this.completedSessions.filter(
      (s) => s.functionName === functionName && s.filePath === filePath
    );

    if (sessions.length === 0) {
      return emptyMetrics(functionName, filePath);
    }

    const durations = sessions
      .filter((s) => s.completedAt !== null)
      .map((s) => s.completedAt! - s.startedAt)
      .sort((a, b) => a - b);

    const nodeTimings = new Map<
      string,
      { total: number; count: number; errors: number }
    >();
    const branchSelections = new Map<string, Map<string, number>>();

    for (const session of sessions) {
      const enterTimes = new Map<string, number>();

      for (const event of session.events) {
        if (event.kind === 'enter') {
          enterTimes.set(event.nodeId, event.timestamp);
        } else if (event.kind === 'exit') {
          const enterTime = enterTimes.get(event.nodeId);
          if (enterTime !== undefined) {
            const duration = event.timestamp - enterTime;
            const existing = nodeTimings.get(event.nodeId) ?? {
              total: 0,
              count: 0,
              errors: 0,
            };
            existing.total += duration;
            existing.count += 1;
            nodeTimings.set(event.nodeId, existing);
          }
        } else if (event.kind === 'error') {
          const existing = nodeTimings.get(event.nodeId) ?? {
            total: 0,
            count: 0,
            errors: 0,
          };
          existing.errors += 1;
          nodeTimings.set(event.nodeId, existing);
        } else if (event.kind === 'fork-branch') {
          const forkId = (event.metadata?.forkNodeId as string) ?? 'unknown';
          const branches =
            branchSelections.get(forkId) ?? new Map<string, number>();
          branches.set(event.nodeId, (branches.get(event.nodeId) ?? 0) + 1);
          branchSelections.set(forkId, branches);
        }
      }
    }

    const totalDuration = durations.reduce((sum, d) => sum + d, 0);
    const hotspots: NodeHotspot[] = [...nodeTimings.entries()]
      .map(([nodeId, timing]) => ({
        nodeId,
        avgDurationMs:
          timing.count > 0 ? Math.round(timing.total / timing.count) : 0,
        invocationCount: timing.count,
        errorCount: timing.errors,
        percentOfTotal:
          totalDuration > 0
            ? Math.round((timing.total / totalDuration) * 100)
            : 0,
      }))
      .sort((a, b) => b.percentOfTotal - a.percentOfTotal);

    const frequencies: BranchFrequency[] = [];
    for (const [forkId, branches] of branchSelections) {
      const totalSelections = [...branches.values()].reduce(
        (sum, c) => sum + c,
        0
      );
      for (const [branchId, count] of branches) {
        frequencies.push({
          forkNodeId: forkId,
          branchNodeId: branchId,
          selectionCount: count,
          selectionRate: totalSelections > 0 ? count / totalSelections : 0,
        });
      }
    }

    // Empirical wallace: based on actual frontier fill during execution
    const maxConcurrency = computeMaxConcurrency(sessions);
    const empiricalWallace = Math.max(
      0,
      1 - maxConcurrency / Math.max(1, hotspots.length)
    );

    return {
      functionName,
      filePath,
      invocationCount: sessions.length,
      avgDurationMs:
        durations.length > 0 ? Math.round(totalDuration / durations.length) : 0,
      p50DurationMs: percentile(durations, 0.5),
      p95DurationMs: percentile(durations, 0.95),
      p99DurationMs: percentile(durations, 0.99),
      nodeHotspots: hotspots,
      branchFrequencies: frequencies,
      empiricalWallace,
      empiricalRegime:
        empiricalWallace < 0.3
          ? 'laminar'
          : empiricalWallace < 0.6
          ? 'transitional'
          : 'turbulent',
    };
  }

  computeWallaceDivergence(
    functionName: string,
    filePath: string,
    staticWallace: number
  ): WallaceDivergence {
    const runtime = this.getMetrics(functionName, filePath);
    const divergence = Math.abs(staticWallace - runtime.empiricalWallace);

    let interpretation: WallaceDivergence['interpretation'];
    if (staticWallace > runtime.empiricalWallace + 0.2) {
      interpretation = 'pessimistic-model';
    } else if (runtime.empiricalWallace > staticWallace + 0.2) {
      interpretation = 'latent-risk';
    } else {
      interpretation = 'aligned';
    }

    return {
      functionName,
      staticWallace,
      empiricalWallace: runtime.empiricalWallace,
      divergence,
      interpretation,
    };
  }

  getActiveSessions(): number {
    return this.sessions.size;
  }

  getCompletedCount(): number {
    return this.completedSessions.length;
  }
}

// ---------------------------------------------------------------------------
// Probe wrapper generator
// ---------------------------------------------------------------------------

/**
 * Creates the Probe Wrapper.
 */
export function createProbeWrapper(
  collector: GnosisProbeCollector,
  functionName: string,
  filePath: string,
  nodeIds: readonly string[]
): <T>(fn: () => Promise<T>) => Promise<T> {
  return async <T>(fn: () => Promise<T>): Promise<T> => {
    const sessionId = collector.startSession(functionName, filePath);

    for (const nodeId of nodeIds) {
      collector.recordEvent(sessionId, { nodeId, kind: 'enter' });
    }

    try {
      const result = await fn();

      for (const nodeId of [...nodeIds].reverse()) {
        collector.recordEvent(sessionId, { nodeId, kind: 'exit' });
      }

      collector.endSession(sessionId);
      return result;
    } catch (error) {
      const lastNode = nodeIds[nodeIds.length - 1] ?? 'unknown';
      collector.recordEvent(sessionId, { nodeId: lastNode, kind: 'error' });
      collector.endSession(sessionId);
      throw error;
    }
  };
}

// ---------------------------------------------------------------------------
// OpenTelemetry span exporter
// ---------------------------------------------------------------------------

export interface OtelSpanAttributes {
  readonly 'gnosis.node.id': string;
  readonly 'gnosis.node.kind': string;
  readonly 'gnosis.function': string;
  readonly 'gnosis.file': string;
  readonly 'gnosis.buley': number;
  readonly 'gnosis.wallace': number;
  readonly 'gnosis.regime': string;
}

/**
 * Builds the Otel Attributes.
 */
export function buildOtelAttributes(
  nodeId: string,
  nodeKind: string,
  functionName: string,
  filePath: string,
  buley: number,
  wallace: number,
  regime: string
): OtelSpanAttributes {
  return {
    'gnosis.node.id': nodeId,
    'gnosis.node.kind': nodeKind,
    'gnosis.function': functionName,
    'gnosis.file': filePath,
    'gnosis.buley': buley,
    'gnosis.wallace': wallace,
    'gnosis.regime': regime,
  };
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

interface MutableSession {
  sessionId: string;
  functionName: string;
  filePath: string;
  startedAt: number;
  events: ProbeEvent[];
  completedAt: number | null;
}

function emptyMetrics(
  functionName: string,
  filePath: string
): RuntimeTopologyMetrics {
  return {
    functionName,
    filePath,
    invocationCount: 0,
    avgDurationMs: 0,
    p50DurationMs: 0,
    p95DurationMs: 0,
    p99DurationMs: 0,
    nodeHotspots: [],
    branchFrequencies: [],
    empiricalWallace: 0,
    empiricalRegime: 'laminar',
  };
}

function percentile(sorted: readonly number[], p: number): number {
  if (sorted.length === 0) return 0;
  const index = Math.ceil(sorted.length * p) - 1;
  const value = sorted[Math.max(0, Math.min(index, sorted.length - 1))];
  return value === undefined ? 0 : value;
}

function computeMaxConcurrency(sessions: readonly MutableSession[]): number {
  let max = 1;
  for (const session of sessions) {
    let concurrent = 0;
    let maxConcurrent = 0;
    for (const event of session.events) {
      if (event.kind === 'enter') concurrent++;
      if (event.kind === 'exit') concurrent--;
      maxConcurrent = Math.max(maxConcurrent, concurrent);
    }
    max = Math.max(max, maxConcurrent);
  }
  return max;
}
