// @a0n/gnosis/src/ts-sonar.ts
import * as fs from 'node:fs';
import * as path from 'node:path';
import * as ts from 'typescript';

export interface TsSonarThresholds {
  maxFileLines?: number;
  maxFunctionLines?: number;
  maxCyclomatic?: number;
  maxCognitive?: number;
  maxNesting?: number;
}

export interface TsFunctionMetric {
  filePath: string;
  name: string;
  startLine: number;
  lineSpan: number;
  cyclomatic: number;
  cognitive: number;
  maxNesting: number;
}

export interface TsFileMetric {
  filePath: string;
  totalLines: number;
  codeLines: number;
  commentLines: number;
  statementCount: number;
  branchCount: number;
  loopCount: number;
  classCount: number;
  interfaceCount: number;
  typeAliasCount: number;
  functionCount: number;
  maxFunctionLines: number;
  maxCyclomatic: number;
  maxCognitive: number;
  maxNesting: number;
}

export interface TsSonarReport {
  ok: boolean;
  fileCount: number;
  totals: {
    totalLines: number;
    codeLines: number;
    commentLines: number;
    statementCount: number;
    branchCount: number;
    loopCount: number;
    classCount: number;
    interfaceCount: number;
    typeAliasCount: number;
    functionCount: number;
    totalCyclomatic: number;
    totalCognitive: number;
    maxFileLines: number;
    maxFunctionLines: number;
    maxCyclomatic: number;
    maxCognitive: number;
    maxNesting: number;
    avgFunctionLines: number;
  };
  files: TsFileMetric[];
  hotspots: {
    byCyclomatic: TsFunctionMetric[];
    byCognitive: TsFunctionMetric[];
    byLength: TsFunctionMetric[];
  };
  violations: string[];
}

const SOURCE_EXTENSIONS = new Set(['.ts', '.tsx', '.js', '.jsx']);
const SKIP_DIR_NAMES = new Set(['node_modules', '.git', 'dist', 'coverage']);

function parseThreshold(
  args: readonly string[],
  flag: string
): number | undefined {
  const index = args.indexOf(flag);
  if (index < 0) {
    return undefined;
  }
  const raw = args[index + 1];
  if (!raw) {
    return undefined;
  }
  const parsed = Number.parseFloat(raw);
  return Number.isFinite(parsed) ? parsed : undefined;
}

/**
 * Parses the Ts Sonar Thresholds.
 */
export function parseTsSonarThresholds(
  args: readonly string[]
): TsSonarThresholds {
  const maxFileLines = parseThreshold(args, '--max-file-lines');
  const maxFunctionLines = parseThreshold(args, '--max-function-lines');
  const maxCyclomatic = parseThreshold(args, '--max-cyclomatic');
  const maxCognitive = parseThreshold(args, '--max-cognitive');
  const maxNesting = parseThreshold(args, '--max-nesting');
  return {
    ...(maxFileLines !== undefined && { maxFileLines }),
    ...(maxFunctionLines !== undefined && { maxFunctionLines }),
    ...(maxCyclomatic !== undefined && { maxCyclomatic }),
    ...(maxCognitive !== undefined && { maxCognitive }),
    ...(maxNesting !== undefined && { maxNesting }),
  };
}

function isSourceFile(filePath: string): boolean {
  const ext = path.extname(filePath).toLowerCase();
  return SOURCE_EXTENSIONS.has(ext);
}

function collectSourceFiles(targetPath: string): string[] {
  if (!fs.existsSync(targetPath)) {
    return [];
  }

  const stat = fs.statSync(targetPath);
  if (stat.isFile()) {
    return isSourceFile(targetPath) ? [targetPath] : [];
  }

  if (!stat.isDirectory()) {
    return [];
  }

  const files: string[] = [];
  const stack: string[] = [targetPath];

  while (stack.length > 0) {
    const current = stack.pop();
    if (!current) {
      continue;
    }
    const entries = fs.readdirSync(current, { withFileTypes: true });
    for (const entry of entries) {
      const resolved = path.join(current, entry.name);
      if (entry.isDirectory()) {
        if (!SKIP_DIR_NAMES.has(entry.name)) {
          stack.push(resolved);
        }
        continue;
      }
      if (entry.isFile() && isSourceFile(resolved)) {
        files.push(resolved);
      }
    }
  }

  files.sort((a, b) => a.localeCompare(b));
  return files;
}

function countCommentAndCodeLines(source: string): {
  codeLines: number;
  commentLines: number;
} {
  const lines = source.split('\n');
  let commentLines = 0;
  let codeLines = 0;
  let inBlockComment = false;

  for (const rawLine of lines) {
    const line = rawLine.trim();
    if (line.length === 0) {
      continue;
    }

    if (inBlockComment) {
      commentLines += 1;
      if (line.includes('*/')) {
        inBlockComment = false;
      }
      continue;
    }

    if (line.startsWith('//')) {
      commentLines += 1;
      continue;
    }

    if (line.startsWith('/*')) {
      commentLines += 1;
      if (!line.includes('*/')) {
        inBlockComment = true;
      }
      continue;
    }

    codeLines += 1;
  }

  return { codeLines, commentLines };
}

function isFunctionLikeWithBody(
  node: ts.Node
): node is
  | ts.FunctionDeclaration
  | ts.MethodDeclaration
  | ts.ArrowFunction
  | ts.FunctionExpression
  | ts.ConstructorDeclaration {
  return (
    (ts.isFunctionDeclaration(node) ||
      ts.isMethodDeclaration(node) ||
      ts.isArrowFunction(node) ||
      ts.isFunctionExpression(node) ||
      ts.isConstructorDeclaration(node)) &&
    Boolean(node.body)
  );
}

function shouldCountDecision(node: ts.Node): boolean {
  return (
    ts.isIfStatement(node) ||
    ts.isForStatement(node) ||
    ts.isForInStatement(node) ||
    ts.isForOfStatement(node) ||
    ts.isWhileStatement(node) ||
    ts.isDoStatement(node) ||
    ts.isCaseClause(node) ||
    ts.isConditionalExpression(node) ||
    ts.isCatchClause(node)
  );
}

function countLogicalDecisionOperators(expression: ts.Expression): number {
  let count = 0;

  function visit(expr: ts.Expression): void {
    if (!ts.isBinaryExpression(expr)) {
      return;
    }
    const operator = expr.operatorToken.kind;
    if (
      operator === ts.SyntaxKind.AmpersandAmpersandToken ||
      operator === ts.SyntaxKind.BarBarToken ||
      operator === ts.SyntaxKind.QuestionQuestionToken
    ) {
      count += 1;
    }
    visit(expr.left);
    visit(expr.right);
  }

  visit(expression);
  return count;
}

function getFunctionName(
  node: TsFunctionLike,
  sourceFile: ts.SourceFile
): string {
  if (
    ts.isFunctionDeclaration(node) ||
    ts.isMethodDeclaration(node) ||
    ts.isFunctionExpression(node)
  ) {
    const name = node.name?.getText(sourceFile);
    if (name) {
      return name;
    }
  }

  if (ts.isConstructorDeclaration(node)) {
    return 'constructor';
  }

  return `<anonymous@${
    sourceFile.getLineAndCharacterOfPosition(node.getStart(sourceFile)).line + 1
  }>`;
}

type TsFunctionLike =
  | ts.FunctionDeclaration
  | ts.MethodDeclaration
  | ts.ArrowFunction
  | ts.FunctionExpression
  | ts.ConstructorDeclaration;

function computeFunctionMetrics(
  node: TsFunctionLike,
  sourceFile: ts.SourceFile,
  filePath: string
): TsFunctionMetric {
  const startLine =
    sourceFile.getLineAndCharacterOfPosition(node.getStart(sourceFile)).line +
    1;
  const endLine =
    sourceFile.getLineAndCharacterOfPosition(node.getEnd()).line + 1;

  let cyclomatic = 1;
  let cognitive = 0;
  let maxNesting = 0;

  const rootBody = node.body;

  if (!rootBody) {
    return {
      filePath,
      name: getFunctionName(node, sourceFile),
      startLine,
      lineSpan: Math.max(1, endLine - startLine + 1),
      cyclomatic,
      cognitive,
      maxNesting,
    };
  }

  function walk(current: ts.Node, nesting: number): void {
    if (current !== rootBody && isFunctionLikeWithBody(current)) {
      return;
    }

    if (shouldCountDecision(current)) {
      cyclomatic += 1;
      cognitive += 1 + nesting;
      maxNesting = Math.max(maxNesting, nesting + 1);
    }

    if (ts.isBinaryExpression(current)) {
      const logicalCount = countLogicalDecisionOperators(current);
      if (logicalCount > 0) {
        cyclomatic += logicalCount;
        cognitive += logicalCount * (1 + nesting);
      }
    }

    const nextNesting =
      shouldCountDecision(current) ||
      ts.isSwitchStatement(current) ||
      ts.isBlock(current)
        ? nesting + 1
        : nesting;

    ts.forEachChild(current, (child) => walk(child, nextNesting));
  }

  walk(rootBody, 0);

  return {
    filePath,
    name: getFunctionName(node, sourceFile),
    startLine,
    lineSpan: Math.max(1, endLine - startLine + 1),
    cyclomatic,
    cognitive,
    maxNesting,
  };
}

function analyzeSourceFile(filePath: string): {
  file: TsFileMetric;
  functions: TsFunctionMetric[];
} {
  // polyglot:ignore RESOURCE_LEAK — readFileSync returns a string, no handle to release
  const source = fs.readFileSync(filePath, 'utf8');
  const sourceFile = ts.createSourceFile(
    filePath,
    source,
    ts.ScriptTarget.Latest,
    true
  );
  const { codeLines, commentLines } = countCommentAndCodeLines(source);
  const totalLines = source.split('\n').length;

  let statementCount = 0;
  let branchCount = 0;
  let loopCount = 0;
  let classCount = 0;
  let interfaceCount = 0;
  let typeAliasCount = 0;
  const functions: TsFunctionMetric[] = [];

  function walk(node: ts.Node): void {
    if (ts.isStatement(node)) {
      statementCount += 1;
    }

    if (shouldCountDecision(node)) {
      branchCount += 1;
    }

    if (
      ts.isForStatement(node) ||
      ts.isForInStatement(node) ||
      ts.isForOfStatement(node) ||
      ts.isWhileStatement(node) ||
      ts.isDoStatement(node)
    ) {
      loopCount += 1;
    }

    if (ts.isClassDeclaration(node)) {
      classCount += 1;
    } else if (ts.isInterfaceDeclaration(node)) {
      interfaceCount += 1;
    } else if (ts.isTypeAliasDeclaration(node)) {
      typeAliasCount += 1;
    }

    if (isFunctionLikeWithBody(node)) {
      functions.push(computeFunctionMetrics(node, sourceFile, filePath));
    }

    ts.forEachChild(node, walk);
  }

  walk(sourceFile);

  const fileMetric: TsFileMetric = {
    filePath,
    totalLines,
    codeLines,
    commentLines,
    statementCount,
    branchCount,
    loopCount,
    classCount,
    interfaceCount,
    typeAliasCount,
    functionCount: functions.length,
    maxFunctionLines: Math.max(
      0,
      ...functions.map((metric) => metric.lineSpan)
    ),
    maxCyclomatic: Math.max(0, ...functions.map((metric) => metric.cyclomatic)),
    maxCognitive: Math.max(0, ...functions.map((metric) => metric.cognitive)),
    maxNesting: Math.max(0, ...functions.map((metric) => metric.maxNesting)),
  };

  return { file: fileMetric, functions };
}

function pushViolations(
  fileMetric: TsFileMetric,
  functions: readonly TsFunctionMetric[],
  thresholds: TsSonarThresholds,
  violations: string[]
): void {
  if (
    thresholds.maxFileLines !== undefined &&
    fileMetric.totalLines > thresholds.maxFileLines
  ) {
    violations.push(
      `file-lines ${fileMetric.filePath} ${fileMetric.totalLines} > ${thresholds.maxFileLines}`
    );
  }

  for (const metric of functions) {
    if (
      thresholds.maxFunctionLines !== undefined &&
      metric.lineSpan > thresholds.maxFunctionLines
    ) {
      violations.push(
        `function-lines ${metric.filePath}:${metric.startLine} ${metric.name} ${metric.lineSpan} > ${thresholds.maxFunctionLines}`
      );
    }
    if (
      thresholds.maxCyclomatic !== undefined &&
      metric.cyclomatic > thresholds.maxCyclomatic
    ) {
      violations.push(
        `cyclomatic ${metric.filePath}:${metric.startLine} ${metric.name} ${metric.cyclomatic} > ${thresholds.maxCyclomatic}`
      );
    }
    if (
      thresholds.maxCognitive !== undefined &&
      metric.cognitive > thresholds.maxCognitive
    ) {
      violations.push(
        `cognitive ${metric.filePath}:${metric.startLine} ${metric.name} ${metric.cognitive} > ${thresholds.maxCognitive}`
      );
    }
    if (
      thresholds.maxNesting !== undefined &&
      metric.maxNesting > thresholds.maxNesting
    ) {
      violations.push(
        `nesting ${metric.filePath}:${metric.startLine} ${metric.name} ${metric.maxNesting} > ${thresholds.maxNesting}`
      );
    }
  }
}

function topBy(
  items: readonly TsFunctionMetric[],
  selector: (item: TsFunctionMetric) => number
): TsFunctionMetric[] {
  return [...items]
    .sort((left, right) => selector(right) - selector(left))
    .slice(0, 10);
}

/**
 * Handles the gnosis analyze Type Script Targets workflow.
 */
export function analyzeTypeScriptTargets(
  targetPaths: readonly string[],
  thresholds: TsSonarThresholds = {}
): TsSonarReport {
  const files = targetPaths.flatMap((targetPath) =>
    collectSourceFiles(targetPath)
  );
  const fileMetrics: TsFileMetric[] = [];
  const functionMetrics: TsFunctionMetric[] = [];
  const violations: string[] = [];

  for (const filePath of files) {
    const { file, functions } = analyzeSourceFile(filePath);
    fileMetrics.push(file);
    functionMetrics.push(...functions);
    pushViolations(file, functions, thresholds, violations);
  }

  const totalLines = fileMetrics.reduce(
    (sum, file) => sum + file.totalLines,
    0
  );
  const codeLines = fileMetrics.reduce((sum, file) => sum + file.codeLines, 0);
  const commentLines = fileMetrics.reduce(
    (sum, file) => sum + file.commentLines,
    0
  );
  const statementCount = fileMetrics.reduce(
    (sum, file) => sum + file.statementCount,
    0
  );
  const branchCount = fileMetrics.reduce(
    (sum, file) => sum + file.branchCount,
    0
  );
  const loopCount = fileMetrics.reduce((sum, file) => sum + file.loopCount, 0);
  const classCount = fileMetrics.reduce(
    (sum, file) => sum + file.classCount,
    0
  );
  const interfaceCount = fileMetrics.reduce(
    (sum, file) => sum + file.interfaceCount,
    0
  );
  const typeAliasCount = fileMetrics.reduce(
    (sum, file) => sum + file.typeAliasCount,
    0
  );
  const functionCount = functionMetrics.length;
  const totalCyclomatic = functionMetrics.reduce(
    (sum, fn) => sum + fn.cyclomatic,
    0
  );
  const totalCognitive = functionMetrics.reduce(
    (sum, fn) => sum + fn.cognitive,
    0
  );

  const report: TsSonarReport = {
    ok: violations.length === 0,
    fileCount: fileMetrics.length,
    totals: {
      totalLines,
      codeLines,
      commentLines,
      statementCount,
      branchCount,
      loopCount,
      classCount,
      interfaceCount,
      typeAliasCount,
      functionCount,
      totalCyclomatic,
      totalCognitive,
      maxFileLines: Math.max(0, ...fileMetrics.map((file) => file.totalLines)),
      maxFunctionLines: Math.max(
        0,
        ...functionMetrics.map((fn) => fn.lineSpan)
      ),
      maxCyclomatic: Math.max(0, ...functionMetrics.map((fn) => fn.cyclomatic)),
      maxCognitive: Math.max(0, ...functionMetrics.map((fn) => fn.cognitive)),
      maxNesting: Math.max(0, ...functionMetrics.map((fn) => fn.maxNesting)),
      avgFunctionLines:
        functionCount === 0
          ? 0
          : Math.round(
              (functionMetrics.reduce((sum, fn) => sum + fn.lineSpan, 0) /
                functionCount) *
                100
            ) / 100,
    },
    files: fileMetrics,
    hotspots: {
      byCyclomatic: topBy(functionMetrics, (fn) => fn.cyclomatic),
      byCognitive: topBy(functionMetrics, (fn) => fn.cognitive),
      byLength: topBy(functionMetrics, (fn) => fn.lineSpan),
    },
    violations,
  };

  return report;
}
