// @a0n/gnosis/src/unscheduling-mesh.ts
/**
 * Unscheduling Mesh — The Dual Runtime Surface
 *
 * The octahedral dual of the tetrahedral scheduling mesh.
 * Where the scheduling mesh does fork/race/fold on active computation,
 * the unscheduling mesh does vent/tunnel/stash on rejected computation.
 *
 * The scheduling mesh is 1/3 of the total computation surface.
 * The unscheduling mesh is the other 2/3.
 *
 * Operations:
 *   VENT   (dual of FORK): route rejected payloads to void pockets
 *   TUNNEL (dual of RACE): navigate between void pockets via shared faces
 *   STASH  (dual of FOLD): aggregate void pocket contents
 *
 * Runtime surfaces:
 *   voidPropagate: BFS through the void edge graph
 *   voidActivate:  full cascade (vent → tunnel → stash → propagate → repeat)
 *   reverseFold:   read from void pocket (costs 1 clinamen)
 *
 * Lean backing:
 *   NegativeSpaceDuality.lean — volume duality, contact faces
 *   VoidStash.lean            — vent-as-write, reverse-fold-as-read
 *   VoidUring.lean            — void ring as speculative prefetch
 *   VoidSIMD.lean             — masked lanes as precomputed void pockets
 *   VoidRuntime.lean          — CPU Model Theorem (savings compound)
 *
 * @packageDocumentation
 */

import { godFormula } from './holy-trinity';
import { type Trit, VENT, VOID, FORK, type MeshVertex } from './scheduling-mesh';

// ─── Void Pocket — The Dual of a Mesh Vertex ───────────────────────

/**
 * A void pocket: a vertex in the unscheduling mesh.
 *
 * Each void pocket is the octahedral dual of a scheduling mesh cell.
 * It stores the rejected payloads, rejection metadata, and tunnel
 * connections to adjacent void pockets.
 *
 * From NegativeSpaceDuality.lean: each scheduling cell face borders
 * exactly one void pocket. The contact surface (Skyrms equilibrium)
 * defines the boundary.
 */
export interface VoidPocket {
  /** Unique pocket ID. */
  readonly id: number;
  /** Corresponding scheduling mesh cell (the dual). */
  readonly dualCellId: number;
  /** Stored payloads (rejected computations). */
  payloads: unknown[];
  /** Total rejections that produced payloads for this pocket. */
  rejectionCount: number;
  /** Reverse-fold weight: w = R - min(v, R) + 1 (inverse of scheduling). */
  weight: number;
  /** How many times this pocket has been read (clinamen counter). */
  reads: number;
  /** Pocket state as a trit. */
  state: Trit;
  /** Timestamp of last write (for FIFO eviction). */
  lastWrite: number;
  /** Timestamp of last read. */
  lastRead: number;
}

/**
 * A void tunnel: an edge in the unscheduling mesh.
 *
 * Tunnels connect adjacent void pockets through shared contact faces.
 * From NegativeSpaceDuality.lean: every face between a tetrahedron
 * and an octahedron is simultaneously a wall of both meshes.
 */
export interface VoidTunnel {
  /** Source pocket ID. */
  readonly from: number;
  /** Target pocket ID. */
  readonly to: number;
  /** Correlation strength (mutual information between pocket contents). */
  correlation: number;
  /** Traversal count (how often this tunnel has been used). */
  traversals: number;
}

// ─── Void Events ────────────────────────────────────────────────────

/** Events emitted by the unscheduling mesh. */
export type VoidEvent =
  | { type: 'vent'; pocketId: number; vertexId: number; hop: number }
  | { type: 'tunnel'; from: number; to: number; correlation: number; hop: number }
  | { type: 'stash'; pocketId: number; payloadCount: number; hop: number }
  | { type: 'reverse-fold'; pocketId: number; clinamen: number; hop: number }
  | { type: 'void-propagate'; from: number; to: number; hops: number; hop: number }
  | { type: 'void-saturated'; hop: number };

/** A firing record in the void activation trace. */
export interface VoidFiring {
  pocketId: number;
  type: 'vent' | 'tunnel' | 'stash' | 'reverse-fold' | 'void-propagate';
  hop: number;
  target?: number;
  clinamen?: number;
}

/** The complete trace of a void activation cascade. */
export interface VoidActivationTrace<T> {
  firings: VoidFiring[];
  results: T[];
  totalHops: number;
  pocketsVisited: number;
  totalClinamen: number;
  startPocket: number;
}

// ─── Vent Record — What the scheduling mesh rejected ────────────────

/**
 * A vent record: the payload and metadata from a rejected computation.
 * This is what the scheduling mesh currently destroys (line 385:
 * `this.vertices[id].payload = null`).
 */
export interface VentRecord {
  /** The rejected vertex ID from the scheduling mesh. */
  readonly vertexId: number;
  /** The rejected payload (would have been nulled). */
  readonly payload: unknown;
  /** The rejection round number. */
  readonly round: number;
  /** God formula weight at time of rejection. */
  readonly weightAtRejection: number;
  /** Timestamp. */
  readonly timestamp: number;
}

// ─── The Unscheduling Mesh ──────────────────────────────────────────

/**
 * UnschedulingMesh — the 2/3 dual of the SchedulingMesh.
 *
 * Where SchedulingMesh operates on vertices with fork/race/fold,
 * UnschedulingMesh operates on void pockets with vent/tunnel/stash.
 *
 * From VoidRuntime.lean: the scheduling mesh is 1/3 of the total
 * computation surface. The unscheduling mesh recovers the other 2/3.
 *
 * From VoidUring.lean: the void ring (pockets) serves as a speculative
 * prefetch cache. Vent is free (byproduct of race). Reverse-fold
 * costs 1 clinamen (observation cost).
 */
export class UnschedulingMesh {
  /** All void pockets. */
  readonly pockets: VoidPocket[];
  /** Adjacency: pocket ID → tunnels. */
  readonly tunnels: Map<number, VoidTunnel[]>;
  /** Total clinamen spent (reads from void). */
  private clinamenSpent: number = 0;
  /** Total vents received (writes to void). */
  private totalVents: number = 0;
  /** God formula rounds (for weight computation). */
  private rounds: number;
  /** Event listeners. */
  private listeners: Array<(event: VoidEvent) => void> = [];
  /** Maximum payloads per pocket before eviction. */
  readonly maxPayloadsPerPocket: number;

  /**
   * Create an unscheduling mesh with one void pocket per scheduling cell.
   *
   * From NegativeSpaceDuality.lean: the void mesh has 2× the volume
   * of the scheduling mesh. We create 2× pockets per scheduling cell
   * to reflect this. For simplicity, dual pockets are paired.
   *
   * @param cellCount - Number of scheduling mesh cells (dual pairs)
   * @param maxPayloadsPerPocket - Max stored payloads before eviction
   */
  constructor(cellCount: number, maxPayloadsPerPocket: number = 64) {
    this.pockets = [];
    this.tunnels = new Map();
    this.rounds = 0;
    this.maxPayloadsPerPocket = maxPayloadsPerPocket;

    // Create 2× pockets (volume duality: void = 2× scheduling)
    const pocketCount = cellCount * 2;
    for (let i = 0; i < pocketCount; i++) {
      this.pockets.push({
        id: i,
        dualCellId: Math.floor(i / 2), // Two pockets per cell
        payloads: [],
        rejectionCount: 0,
        weight: 1,
        reads: 0,
        state: VOID,
        lastWrite: 0,
        lastRead: 0,
      });
      this.tunnels.set(i, []);
    }

    // Connect pockets via tunnels (contact face adjacency)
    this.buildTunnels(pocketCount);
  }

  /**
   * Build tunnel connections between void pockets.
   *
   * Tunnel topology mirrors the octahedral face adjacency:
   * - Paired pockets (same cell) are always connected
   * - Adjacent-cell pockets are connected (face-sharing)
   */
  private buildTunnels(pocketCount: number): void {
    for (let i = 0; i < pocketCount; i++) {
      // Paired pocket connection (two pockets per cell)
      const paired = i % 2 === 0 ? i + 1 : i - 1;
      if (paired >= 0 && paired < pocketCount) {
        this.addTunnel(i, paired);
      }

      // Adjacent-cell connections
      // Each pocket connects to pockets in neighboring cells
      const cellId = Math.floor(i / 2);
      const nextCell = cellId + 1;
      const prevCell = cellId - 1;

      if (nextCell * 2 + 1 < pocketCount) {
        this.addTunnel(i, nextCell * 2);
        this.addTunnel(i, nextCell * 2 + 1);
      }
      if (prevCell >= 0) {
        this.addTunnel(i, prevCell * 2);
        this.addTunnel(i, prevCell * 2 + 1);
      }
    }
  }

  private addTunnel(from: number, to: number): void {
    if (from === to) return;
    const existing = this.tunnels.get(from);
    if (!existing) return;
    // Avoid duplicates
    if (existing.some(t => t.to === to)) return;
    existing.push({ from, to, correlation: 0, traversals: 0 });
  }

  // ─── Core Operations ───────────────────────────────────────────

  /**
   * VENT: Route a rejected payload to a void pocket.
   *
   * The dual of FORK. Instead of splitting work, you store rejected work.
   * From VoidStash.lean: vent_write_monotone — stash grows monotonically.
   * From VoidUring.lean: encode_is_free — venting costs nothing.
   *
   * @param record - The vent record (rejected vertex payload)
   * @returns The pocket ID where the payload was stored
   */
  vent(record: VentRecord): number {
    const pocketId = this.selectPocket(record.vertexId);
    const pocket = this.pockets[pocketId];
    if (pocket === undefined) {
      throw new Error(`Missing void pocket ${pocketId}`);
    }

    // Evict oldest if at capacity
    if (pocket.payloads.length >= this.maxPayloadsPerPocket) {
      pocket.payloads.shift(); // FIFO eviction
    }

    pocket.payloads.push(record);
    pocket.rejectionCount++;
    pocket.lastWrite = Date.now();
    pocket.state = VENT; // Pocket has content
    this.totalVents++;

    // Update weight: inverse of scheduling weight
    // High rejection count = high void weight (most data available)
    this.rounds++;
    pocket.weight = godFormula(this.rounds, 0) - godFormula(this.rounds, record.weightAtRejection) + 1;

    return pocketId;
  }

  /**
   * Select which pocket receives a vent. Round-robins across the
   * two pockets for the cell that produced the rejection.
   */
  private selectPocket(vertexId: number): number {
    // Map vertex to cell (simplified: vertexId mod pocketCount/2)
    const cellId = vertexId % (this.pockets.length / 2);
    const pocket0 = cellId * 2;
    const pocket1 = pocket0 + 1;

    // Choose the pocket with fewer payloads (load balance)
    const p0 = this.pockets[pocket0];
    const p1 = this.pockets[pocket1];
    if (p0 === undefined || p1 === undefined) {
      throw new Error(`Missing void pocket pair ${pocket0}/${pocket1}`);
    }
    return p0.payloads.length <= p1.payloads.length ? pocket0 : pocket1;
  }

  /**
   * TUNNEL: Navigate between void pockets via shared faces.
   *
   * The dual of RACE. Instead of competing paths, you correlate
   * rejection records. Returns pockets connected to the source
   * with matching vent records.
   *
   * @param sourcePocket - Starting pocket ID
   * @param predicate - Filter for relevant vent records
   * @returns Array of matching pocket IDs with their correlation
   */
  tunnel(
    sourcePocket: number,
    predicate: (record: VentRecord) => boolean,
  ): Array<{ pocketId: number; correlation: number; matches: number }> {
    const results: Array<{ pocketId: number; correlation: number; matches: number }> = [];
    const tunnelEdges = this.tunnels.get(sourcePocket) ?? [];

    for (const tunnel of tunnelEdges) {
      const targetPocket = this.pockets[tunnel.to];
      if (targetPocket === undefined) {
        continue;
      }
      const matches = targetPocket.payloads.filter(
        p => predicate(p as VentRecord),
      ).length;

      if (matches > 0) {
        // Update tunnel correlation based on match density
        const density = matches / Math.max(1, targetPocket.payloads.length);
        tunnel.correlation = density;
        tunnel.traversals++;

        results.push({
          pocketId: tunnel.to,
          correlation: density,
          matches,
        });
      }
    }

    // Sort by correlation (highest first — most relevant void pocket)
    results.sort((a, b) => b.correlation - a.correlation);
    return results;
  }

  /**
   * STASH: Aggregate void pocket contents across tunnels.
   *
   * The dual of FOLD. Instead of merging winners, you accumulate
   * rejection records from connected pockets.
   *
   * @param sourcePocket - Hub pocket ID
   * @param predicate - Filter for relevant records
   * @returns Aggregated vent records from all tunnel-connected pockets
   */
  stash(
    sourcePocket: number,
    predicate: (record: VentRecord) => boolean,
  ): VentRecord[] {
    const aggregated: VentRecord[] = [];

    // Include source pocket's matching records
    const source = this.pockets[sourcePocket];
    if (source === undefined) {
      return aggregated;
    }
    for (const p of source.payloads) {
      if (predicate(p as VentRecord)) {
        aggregated.push(p as VentRecord);
      }
    }

    // Collect from tunnel-connected pockets
    const tunnelEdges = this.tunnels.get(sourcePocket) ?? [];
    for (const tunnel of tunnelEdges) {
      const target = this.pockets[tunnel.to];
      if (target === undefined) {
        continue;
      }
      for (const p of target.payloads) {
        if (predicate(p as VentRecord)) {
          aggregated.push(p as VentRecord);
        }
      }
    }

    return aggregated;
  }

  /**
   * REVERSE FOLD: Read from a void pocket.
   *
   * From VoidStash.lean: read_cost_is_clinamen — each read costs 1.
   * From VoidCodec.lean: decode_costs_clinamen — observation cost.
   *
   * This is the only operation that costs clinamen. Everything else
   * (vent, tunnel, stash) is free.
   *
   * @param pocketId - The pocket to read from
   * @returns All vent records in the pocket, or null if empty
   */
  reverseFold(pocketId: number): VentRecord[] | null {
    const pocket = this.pockets[pocketId];
    if (pocket === undefined) {
      return null;
    }
    if (pocket.payloads.length === 0) return null;

    // Costs 1 clinamen
    this.clinamenSpent++;
    pocket.reads++;
    pocket.lastRead = Date.now();

    return pocket.payloads.slice() as VentRecord[];
  }

  // ─── Traversal ─────────────────────────────────────────────────

  /**
   * BFS propagation through the void mesh.
   *
   * The dual of SchedulingMesh.propagate(). Walks void tunnels
   * instead of scheduling edges.
   */
  voidPropagate(
    sourcePocket: number,
    visitor: (pocket: VoidPocket, hop: number) => boolean,
  ): number {
    const visited = new Set<number>();
    const queue: Array<{ id: number; hop: number }> = [{ id: sourcePocket, hop: 0 }];
    visited.add(sourcePocket);
    let reached = 0;

    while (queue.length > 0) {
      const { id, hop } = queue.shift()!;
      const pocket = this.pockets[id];
      if (pocket === undefined) {
        continue;
      }

      const shouldContinue = visitor(pocket, hop);
      reached++;

      if (!shouldContinue) continue;

      const tunnelEdges = this.tunnels.get(id) ?? [];
      for (const tunnel of tunnelEdges) {
        if (!visited.has(tunnel.to)) {
          visited.add(tunnel.to);
          tunnel.traversals++;
          queue.push({ id: tunnel.to, hop: hop + 1 });
        }
      }
    }

    return reached;
  }

  /**
   * BFS shortest path between two void pockets.
   */
  voidPath(from: number, to: number): number[] | null {
    if (from === to) return [from];

    const visited = new Set<number>();
    const parent = new Map<number, number>();
    const queue: number[] = [from];
    visited.add(from);

    while (queue.length > 0) {
      const current = queue.shift()!;
      const tunnelEdges = this.tunnels.get(current) ?? [];

      for (const tunnel of tunnelEdges) {
        if (!visited.has(tunnel.to)) {
          visited.add(tunnel.to);
          parent.set(tunnel.to, current);

          if (tunnel.to === to) {
            const path: number[] = [to];
            let node = to;
            while (parent.has(node)) {
              node = parent.get(node)!;
              path.unshift(node);
            }
            return path;
          }

          queue.push(tunnel.to);
        }
      }
    }

    return null;
  }

  // ─── Cascade ───────────────────────────────────────────────────

  /**
   * Activate the void mesh — full vent/tunnel/stash cascade.
   *
   * The dual of SchedulingMesh.activate().
   *
   * 1. Start at a void pocket
   * 2. TUNNEL: find connected pockets with correlated content
   * 3. STASH: aggregate matching records
   * 4. Extract results from aggregated records
   * 5. PROPAGATE: pathfind to the highest-content pocket
   * 6. REPEAT
   *
   * The cascade stops when:
   * - No more pockets with matching content
   * - maxHops exceeded
   * - The process function returns null
   */
  voidActivate<T>(
    startPocket: number,
    predicate: (record: VentRecord) => boolean,
    process: (records: VentRecord[], pocketId: number) => T | null,
    maxHops: number = this.pockets.length,
  ): VoidActivationTrace<T> {
    const trace: VoidActivationTrace<T> = {
      firings: [],
      results: [],
      totalHops: 0,
      pocketsVisited: 0,
      totalClinamen: 0,
      startPocket,
    };

    let currentPocket = startPocket;

    for (let hop = 0; hop < maxHops; hop++) {
      // 1. REVERSE FOLD: read current pocket (costs 1 clinamen)
      const records = this.reverseFold(currentPocket);
      if (!records || records.length === 0) break;

      trace.totalClinamen++;
      this.emit({ type: 'reverse-fold', pocketId: currentPocket, clinamen: this.clinamenSpent, hop });
      trace.firings.push({
        pocketId: currentPocket,
        type: 'reverse-fold',
        hop,
        clinamen: this.clinamenSpent,
      });

      // 2. TUNNEL: find correlated pockets
      const tunnelResults = this.tunnel(currentPocket, predicate);
      if (tunnelResults.length > 0) {
        const best = tunnelResults[0];
        if (best === undefined) {
          break;
        }
        this.emit({
          type: 'tunnel',
          from: currentPocket,
          to: best.pocketId,
          correlation: best.correlation,
          hop,
        });
        trace.firings.push({
          pocketId: currentPocket,
          type: 'tunnel',
          hop,
          target: best.pocketId,
        });
      }

      // 3. STASH: aggregate matching records
      const stashed = this.stash(currentPocket, predicate);
      this.emit({ type: 'stash', pocketId: currentPocket, payloadCount: stashed.length, hop });
      trace.firings.push({
        pocketId: currentPocket,
        type: 'stash',
        hop,
      });

      // 4. Process aggregated records
      const result = process(stashed, currentPocket);
      if (result === null) break;

      trace.results.push(result);
      trace.pocketsVisited++;

      // 5. PROPAGATE: pathfind to the highest-content connected pocket
      const nextPocket = this.highestContentPocket(currentPocket);
      if (nextPocket === null) {
        this.emit({ type: 'void-saturated', hop });
        break;
      }

      this.emit({
        type: 'void-propagate',
        from: currentPocket,
        to: nextPocket,
        hops: 1,
        hop,
      });
      trace.firings.push({
        pocketId: currentPocket,
        type: 'void-propagate',
        hop,
        target: nextPocket,
      });
      trace.totalHops++;

      currentPocket = nextPocket;
    }

    return trace;
  }

  /**
   * Find the connected pocket with the most content (highest payload count).
   * The void mesh equivalent of "nearest idle hub."
   */
  private highestContentPocket(from: number): number | null {
    const tunnelEdges = this.tunnels.get(from) ?? [];
    let bestPocket: number | null = null;
    let bestContent = 0;

    for (const tunnel of tunnelEdges) {
      const pocket = this.pockets[tunnel.to];
      if (pocket === undefined) {
        continue;
      }
      if (pocket.payloads.length > bestContent) {
        bestContent = pocket.payloads.length;
        bestPocket = tunnel.to;
      }
    }

    return bestPocket;
  }

  // ─── Void Ring (io_uring Speculative Prefetch) ─────────────────

  /**
   * Check the void ring for a cached result before submitting
   * new work. This is the speculative prefetch from VoidUring.lean.
   *
   * Returns matching vent records without costing clinamen
   * (index scan only — scheduling mesh lookup).
   */
  checkVoidRing(
    predicate: (record: VentRecord) => boolean,
  ): VentRecord | null {
    // Scan all pockets for matching content (O(n) — should be indexed)
    for (const pocket of this.pockets) {
      for (const p of pocket.payloads) {
        if (predicate(p as VentRecord)) {
          return p as VentRecord;
        }
      }
    }
    return null;
  }

  /**
   * Speculative check + read. If a match is found, read it (1 clinamen).
   * If not, return null (0 cost — the miss is free).
   *
   * From VoidUring.lean: void_read_cheaper — checking the void ring
   * before submitting a new SQE saves a ring transition.
   */
  speculativeRead(
    predicate: (record: VentRecord) => boolean,
  ): VentRecord | null {
    const match = this.checkVoidRing(predicate);
    if (match) {
      this.clinamenSpent++; // reading costs 1 clinamen
    }
    return match;
  }

  // ─── Event System ──────────────────────────────────────────────

  /** Subscribe to void mesh events. */
  on(listener: (event: VoidEvent) => void): void {
    this.listeners.push(listener);
  }

  private emit(event: VoidEvent): void {
    for (const fn of this.listeners) fn(event);
  }

  // ─── Queries ───────────────────────────────────────────────────

  /** Get all non-empty pockets. */
  activePockets(): number[] {
    return this.pockets.filter(p => p.payloads.length > 0).map(p => p.id);
  }

  /** Get all empty pockets. */
  emptyPockets(): number[] {
    return this.pockets.filter(p => p.payloads.length === 0).map(p => p.id);
  }

  /** Total payloads stored across all pockets. */
  get totalPayloads(): number {
    return this.pockets.reduce((sum, p) => sum + p.payloads.length, 0);
  }

  /** Void mesh statistics. */
  stats(): {
    pockets: number;
    tunnelCount: number;
    totalPayloads: number;
    activePockets: number;
    emptyPockets: number;
    totalVents: number;
    clinamenSpent: number;
    preservedWork: number;
    hitRate: number;
    averagePayloadsPerPocket: number;
    highestCorrelationTunnel: { from: number; to: number; correlation: number } | null;
  } {
    let tunnelCount = 0;
    let highestCorrelation: VoidTunnel | null = null;

    for (const edges of this.tunnels.values()) {
      tunnelCount += edges.length;
      for (const tunnel of edges) {
        if (!highestCorrelation || tunnel.correlation > highestCorrelation.correlation) {
          highestCorrelation = tunnel;
        }
      }
    }
    tunnelCount = Math.floor(tunnelCount / 2); // Bidirectional

    const activePocketCount = this.activePockets().length;
    const totalPayloads = this.totalPayloads;

    return {
      pockets: this.pockets.length,
      tunnelCount,
      totalPayloads,
      activePockets: activePocketCount,
      emptyPockets: this.pockets.length - activePocketCount,
      totalVents: this.totalVents,
      clinamenSpent: this.clinamenSpent,
      preservedWork: this.totalVents, // Each vent preserves 1 unit of work
      hitRate: this.totalVents > 0
        ? this.clinamenSpent / this.totalVents
        : 0,
      averagePayloadsPerPocket: this.pockets.length > 0
        ? totalPayloads / this.pockets.length
        : 0,
      highestCorrelationTunnel: highestCorrelation
        ? {
            from: highestCorrelation.from,
            to: highestCorrelation.to,
            correlation: highestCorrelation.correlation,
          }
        : null,
    };
  }

  /** Reset the entire void mesh. */
  reset(): void {
    for (const pocket of this.pockets) {
      pocket.payloads = [];
      pocket.rejectionCount = 0;
      pocket.weight = 1;
      pocket.reads = 0;
      pocket.state = VOID;
      pocket.lastWrite = 0;
      pocket.lastRead = 0;
    }
    for (const edges of this.tunnels.values()) {
      for (const tunnel of edges) {
        tunnel.correlation = 0;
        tunnel.traversals = 0;
      }
    }
    this.clinamenSpent = 0;
    this.totalVents = 0;
    this.rounds = 0;
  }
}

// ─── Factory ──────────────────────────────────────────────────────

/**
 * Create an unscheduling mesh sized to match a scheduling mesh.
 *
 * @param schedulingCells - Number of cells in the corresponding scheduling mesh
 * @param maxPayloadsPerPocket - Maximum payloads per void pocket
 */
export function createUnschedulingMesh(
  schedulingCells: number,
  maxPayloadsPerPocket: number = 64,
): UnschedulingMesh {
  return new UnschedulingMesh(schedulingCells, maxPayloadsPerPocket);
}

// ─── Dual Mesh — Combined Scheduling + Unscheduling ─────────────

/**
 * DualMesh: the scheduling mesh and its unscheduling dual, combined.
 *
 * This is the full 3/3 computation surface. The scheduling mesh
 * handles active work (1/3). The unscheduling mesh handles rejected
 * work (2/3). Together, they capture all computation the hardware
 * performs.
 *
 * From VoidRuntime.lean: savings compound vertically across layers.
 * topology × uring × SIMD = maximum recovery.
 */
export interface DualMeshStats {
  scheduling: {
    vertices: number;
    edges: number;
    cells: number;
    activeVertices: number;
    ventedVertices: number;
    totalRejections: number;
  };
  unscheduling: {
    pockets: number;
    tunnels: number;
    totalPayloads: number;
    activePockets: number;
    clinamenSpent: number;
    preservedWork: number;
  };
  combined: {
    /** Total computation surface (scheduling + unscheduling). */
    totalSurface: number;
    /** Fraction of surface used by scheduling (should be ~1/3). */
    schedulingFraction: number;
    /** Fraction of surface used by unscheduling (should be ~2/3). */
    unschedulingFraction: number;
    /** Work preserved by the unscheduling mesh. */
    workPreserved: number;
    /** Clinamen efficiency: preserved / clinamen. */
    clinamenEfficiency: number;
  };
}

/**
 * Compute combined stats for a scheduling + unscheduling mesh pair.
 */
export function dualMeshStats(
  schedulingStats: {
    vertices: number;
    edges: number;
    cells: number;
    activeVertices: number;
    ventedVertices: number;
    totalRejections: number;
  },
  unschedulingStats: ReturnType<UnschedulingMesh['stats']>,
): DualMeshStats {
  const totalSurface = schedulingStats.vertices + unschedulingStats.pockets;
  const schedulingFraction = totalSurface > 0
    ? schedulingStats.vertices / totalSurface
    : 0;
  const unschedulingFraction = totalSurface > 0
    ? unschedulingStats.pockets / totalSurface
    : 0;

  return {
    scheduling: schedulingStats,
    unscheduling: {
      pockets: unschedulingStats.pockets,
      tunnels: unschedulingStats.tunnelCount,
      totalPayloads: unschedulingStats.totalPayloads,
      activePockets: unschedulingStats.activePockets,
      clinamenSpent: unschedulingStats.clinamenSpent,
      preservedWork: unschedulingStats.preservedWork,
    },
    combined: {
      totalSurface,
      schedulingFraction,
      unschedulingFraction,
      workPreserved: unschedulingStats.preservedWork,
      clinamenEfficiency: unschedulingStats.clinamenSpent > 0
        ? unschedulingStats.preservedWork / unschedulingStats.clinamenSpent
        : Infinity,
    },
  };
}
