// @a0n/gnosis/src/unspeech-triton.ts
/**
 * Unspeech Triton -- Signed Timing Channel
 *
 * TypeScript runtime mirror of ForkRaceFoldTheorems/UnspeechTriton.lean.
 *
 * Core ideas:
 * 1. Triton alphabet {-1, 0, +1} keeps directional timing signal.
 * 2. Gap quantization maps continuous timing into discrete unspeech symbols.
 * 3. Stochastic channel kernel estimates expected repair burden.
 * 4. English carrier surface embeds trits in syntactically valid variants.
 */
import { cacheFingerprint64HexFrfSync } from '@a0n/bitwise/cache-fp64';

// ---------------------------------------------------------------------------
// Part I: Core alphabet
// ---------------------------------------------------------------------------

export type Triton = -1 | 0 | 1;

export const TRITON_SYMBOLS: readonly Triton[] = [-1, 0, 1] as const;

/** Binary collapse loses sign direction. */
export function binaryCollapse(t: Triton): 0 | 1 {
  return t === 0 ? 0 : 1;
}

/** Witness: binary collapse cannot separate -1 from +1. */
export function binaryCollisionWitness(): boolean {
  return binaryCollapse(-1) === binaryCollapse(1);
}

// ---------------------------------------------------------------------------
// Part II: Gap quantization
// ---------------------------------------------------------------------------

/**
 * Quantize a measured gap around baseline with tolerance epsilon.
 * -1: compressed gap
 *  0: neutral band
 * +1: expanded gap
 */
export function quantizeGap(
  baseline: number,
  epsilon: number,
  measuredGap: number
): Triton {
  if (measuredGap + epsilon < baseline) return -1;
  if (baseline + epsilon < measuredGap) return 1;
  return 0;
}

// ---------------------------------------------------------------------------
// Part III: Discrete stochastic channel
// ---------------------------------------------------------------------------

export type TritonLane = 'neg' | 'zero' | 'pos';

export interface TritonMassVector {
  neg: number;
  zero: number;
  pos: number;
}

export interface TritonMassKernel {
  neg: TritonMassVector;
  zero: TritonMassVector;
  pos: TritonMassVector;
}

/**
 * Handles the gnosis triton To Lane workflow.
 */
export function tritonToLane(t: Triton): TritonLane {
  if (t === -1) return 'neg';
  if (t === 1) return 'pos';
  return 'zero';
}

/**
 * Handles the gnosis lane To Triton workflow.
 */
export function laneToTriton(lane: TritonLane): Triton {
  if (lane === 'neg') return -1;
  if (lane === 'pos') return 1;
  return 0;
}

/** Unit correction mass: mismatch incurs one correction unit. */
export function symbolCorrectionCost(sent: Triton, recv: Triton): 0 | 1 {
  return sent === recv ? 0 : 1;
}

/**
 * Handles the gnosis row Mass workflow.
 */
export function rowMass(kernel: TritonMassKernel, sent: Triton): number {
  const row = kernel[tritonToLane(sent)];
  return row.neg + row.zero + row.pos;
}

/**
 * Handles the gnosis correction Mass workflow.
 */
export function correctionMass(kernel: TritonMassKernel, sent: Triton): number {
  const row = kernel[tritonToLane(sent)];
  return (
    row.neg * symbolCorrectionCost(sent, -1) +
    row.zero * symbolCorrectionCost(sent, 0) +
    row.pos * symbolCorrectionCost(sent, 1)
  );
}

/**
 * Handles the gnosis no Confusion For workflow.
 */
export function noConfusionFor(kernel: TritonMassKernel, sent: Triton): boolean {
  const row = kernel[tritonToLane(sent)];
  if (sent === -1) return row.zero === 0 && row.pos === 0;
  if (sent === 0) return row.neg === 0 && row.pos === 0;
  return row.neg === 0 && row.zero === 0;
}

/**
 * Handles the gnosis normalized Row workflow.
 */
export function normalizedRow(
  kernel: TritonMassKernel,
  sent: Triton
): { neg: number; zero: number; pos: number } {
  const row = kernel[tritonToLane(sent)];
  const mass = rowMass(kernel, sent);
  if (mass <= 0) return { neg: 0, zero: 0, pos: 0 };
  return {
    neg: row.neg / mass,
    zero: row.zero / mass,
    pos: row.pos / mass,
  };
}

/**
 * Handles the gnosis expected Correction Rate workflow.
 */
export function expectedCorrectionRate(
  kernel: TritonMassKernel,
  sent: Triton
): number {
  const mass = rowMass(kernel, sent);
  if (mass <= 0) return 0;
  return correctionMass(kernel, sent) / mass;
}

// ---------------------------------------------------------------------------
// Part IV: Objective surface
// ---------------------------------------------------------------------------

export interface UnspeechPolicyCost {
  semanticError: number;
  repairCost: number;
  articulationCost: number;
  latencyCost: number;
}

/**
 * Handles the gnosis unspeech Objective workflow.
 */
export function unspeechObjective(
  lambda: number,
  mu: number,
  nu: number,
  cost: UnspeechPolicyCost
): number {
  return (
    cost.semanticError +
    lambda * cost.repairCost +
    mu * cost.articulationCost +
    nu * cost.latencyCost
  );
}

/**
 * Handles the gnosis prefers Lower Repair When Others Equal workflow.
 */
export function prefersLowerRepairWhenOthersEqual(
  lambda: number,
  a: UnspeechPolicyCost,
  b: UnspeechPolicyCost
): boolean {
  const sameOtherTerms =
    a.semanticError === b.semanticError &&
    a.articulationCost === b.articulationCost &&
    a.latencyCost === b.latencyCost;
  if (!sameOtherTerms) return false;
  if (!(a.repairCost < b.repairCost)) return false;
  return unspeechObjective(lambda, 1, 1, a) < unspeechObjective(lambda, 1, 1, b);
}

// ---------------------------------------------------------------------------
// Part V: English carrier (syntactically valid)
// ---------------------------------------------------------------------------

export type ThatVariant = 'dropped' | 'explicit' | 'emphatic';

/**
 * Handles the gnosis encode That workflow.
 */
export function encodeThat(t: Triton): ThatVariant {
  if (t === -1) return 'dropped';
  if (t === 1) return 'emphatic';
  return 'explicit';
}

/**
 * Handles the gnosis decode That workflow.
 */
export function decodeThat(v: ThatVariant): Triton {
  if (v === 'dropped') return -1;
  if (v === 'emphatic') return 1;
  return 0;
}

/**
 * Handles the gnosis encode That Sequence workflow.
 */
export function encodeThatSequence(payload: readonly Triton[]): ThatVariant[] {
  return payload.map(encodeThat);
}

/**
 * Handles the gnosis decode That Sequence workflow.
 */
export function decodeThatSequence(payload: readonly ThatVariant[]): Triton[] {
  return payload.map(decodeThat);
}

function trimFinalPunctuation(input: string): string {
  return input.trim().replace(/[.!?]+$/u, '').trim();
}

/** Render one triton into a syntactically valid English sentence. */
export function renderThatCarrier(t: Triton, clause: string): string {
  const normalized = trimFinalPunctuation(clause);
  if (t === -1) return `I think ${normalized}.`;
  if (t === 1) return `I do think that ${normalized}.`;
  return `I think that ${normalized}.`;
}

/**
 * Decode triton from carrier form.
 * Returns null when sentence is outside supported carrier templates.
 */
export function decodeThatCarrierSentence(sentence: string): Triton | null {
  const normalized = sentence.trim().toLowerCase().replace(/\s+/gu, ' ');
  if (normalized.startsWith('i do think that ')) return 1;
  if (normalized.startsWith('i think that ')) return 0;
  if (normalized.startsWith('i think ')) return -1;
  return null;
}

// ---------------------------------------------------------------------------
// Part VI: Stochastic transmission simulator
// ---------------------------------------------------------------------------

export type TritonRng = () => number;

/**
 * Handles the gnosis sample Received workflow.
 */
export function sampleReceived(
  kernel: TritonMassKernel,
  sent: Triton,
  rng: TritonRng = Math.random
): Triton {
  const row = kernel[tritonToLane(sent)];
  const total = row.neg + row.zero + row.pos;
  if (total <= 0) return sent;

  const draw = rng() * total;
  const first = row.neg;
  const second = row.neg + row.zero;

  if (draw < first) return -1;
  if (draw < second) return 0;
  return 1;
}

export interface TransmissionResult {
  sent: Triton[];
  received: Triton[];
  corrections: number;
}

/**
 * Handles the gnosis transmit Sequence workflow.
 */
export function transmitSequence(
  payload: readonly Triton[],
  kernel: TritonMassKernel,
  rng: TritonRng = Math.random
): TransmissionResult {
  const sent = [...payload];
  const received = sent.map((symbol) => sampleReceived(kernel, symbol, rng));
  const corrections = sent.reduce<number>((acc, symbol, idx) => {
    return acc + symbolCorrectionCost(symbol, received[idx] as Triton);
  }, 0);
  return { sent, received, corrections };
}

// ---------------------------------------------------------------------------
// Part VII: Self-similar triton compression (composable with bitwise)
// ---------------------------------------------------------------------------

const SELF_SIMILAR_MAGIC = new Uint8Array([0x55, 0x54, 0x53, 0x52]); // UTSR
const SELF_SIMILAR_VERSION = 1;
const SELF_SIMILAR_HEADER_SIZE = 14;

export type TritonSelfSimilarEncoding =
  | 'self-similar-rle'
  | 'self-similar-rle+bitwise-fold';

export interface TritonSelfSimilarCompressionOptions {
  motifSize?: number;
  useBitwiseFold?: boolean;
  preferSmallerPayload?: boolean;
}

export interface TritonSelfSimilarPacket {
  encoding: TritonSelfSimilarEncoding;
  motifSize: number;
  tritonLength: number;
  rawLength: number;
  coreLength: number;
  payloadLength: number;
  compressionRatio: number;
  coreFingerprintHex: string;
  payloadFingerprintHex: string;
  payload: Uint8Array;
}

export interface TritonSelfSimilarBatch {
  packets: TritonSelfSimilarPacket[];
  totalRawLength: number;
  totalCompressedLength: number;
  compressionRatio: number;
  aggregateFingerprintHex: string;
}

interface VarUintRead {
  value: number;
  nextOffset: number;
}

interface BitwisePkgModule {
  syzygy_fold_binary: (payload: Uint8Array) => Uint8Array;
  syzygy_unfold_binary: (payload: Uint8Array, originalLength: number) => Uint8Array;
}

function ensureMotifSize(motifSize: number): number {
  if (!Number.isInteger(motifSize) || motifSize < 1 || motifSize > 5) {
    throw new RangeError(
      `motifSize must be an integer in [1, 5], received ${motifSize}`
    );
  }
  return motifSize;
}

function writeU32BE(output: Uint8Array, offset: number, value: number): void {
  if (!Number.isInteger(value) || value < 0 || value > 0xffffffff) {
    throw new RangeError(`u32 overflow at offset ${offset}: ${value}`);
  }
  output[offset] = Math.floor(value / 0x1000000) & 0xff;
  output[offset + 1] = Math.floor(value / 0x10000) & 0xff;
  output[offset + 2] = Math.floor(value / 0x100) & 0xff;
  output[offset + 3] = value & 0xff;
}

function readU32BE(input: Uint8Array, offset: number): number {
  return (
    input[offset]! * 0x1000000 +
    input[offset + 1]! * 0x10000 +
    input[offset + 2]! * 0x100 +
    input[offset + 3]!
  );
}

function writeVarUint(output: number[], value: number): void {
  if (!Number.isInteger(value) || value < 0) {
    throw new RangeError(`varuint must be a non-negative integer, received ${value}`);
  }
  let remainder = value;
  do {
    let byte = remainder % 128;
    remainder = Math.floor(remainder / 128);
    if (remainder > 0) byte += 128;
    output.push(byte);
  } while (remainder > 0);
}

function readVarUint(input: Uint8Array, startOffset: number): VarUintRead {
  let value = 0;
  let factor = 1;
  let offset = startOffset;
  for (let i = 0; i < 8; i++) {
    if (offset >= input.length) {
      throw new RangeError('unexpected end of payload while reading varuint');
    }
    const byte = input[offset]!;
    offset += 1;
    value += (byte & 0x7f) * factor;
    if ((byte & 0x80) === 0) {
      return { value, nextOffset: offset };
    }
    factor *= 128;
  }
  throw new RangeError('varuint exceeds supported length');
}

function tritonToDigit(t: Triton): 0 | 1 | 2 {
  if (t === -1) return 0;
  if (t === 0) return 1;
  return 2;
}

function digitToTriton(d: number): Triton {
  if (d === 0) return -1;
  if (d === 1) return 0;
  if (d === 2) return 1;
  throw new RangeError(`invalid triton digit ${d}`);
}

function pow3(exp: number): number {
  let out = 1;
  for (let i = 0; i < exp; i++) out *= 3;
  return out;
}

function encodeMotifByte(
  payload: readonly Triton[],
  start: number,
  motifSize: number
): number {
  let value = 0;
  for (let i = 0; i < motifSize; i++) {
    const symbol = start + i < payload.length ? payload[start + i]! : 0;
    value = value * 3 + tritonToDigit(symbol);
  }
  return value;
}

function decodeMotifByte(value: number, motifSize: number): Triton[] {
  let remainder = value;
  const motif = new Array<Triton>(motifSize).fill(0);
  for (let i = motifSize - 1; i >= 0; i--) {
    motif[i] = digitToTriton(remainder % 3);
    remainder = Math.floor(remainder / 3);
  }
  return motif;
}

function concatUint8Arrays(chunks: readonly Uint8Array[]): Uint8Array {
  const total = chunks.reduce((sum, chunk) => sum + chunk.length, 0);
  const out = new Uint8Array(total);
  let offset = 0;
  for (const chunk of chunks) {
    out.set(chunk, offset);
    offset += chunk.length;
  }
  return out;
}

async function loadBitwisePkgModule(): Promise<BitwisePkgModule> {
  return (await import('@a0n/bitwise/pkg')) as unknown as BitwisePkgModule;
}

async function tryFoldWithBitwiseBinary(
  payload: Uint8Array
): Promise<Uint8Array | null> {
  try {
    const bitwise = await loadBitwisePkgModule();
    return bitwise.syzygy_fold_binary(payload);
  } catch {
    return null;
  }
}

/**
 * Encode triton stream into self-similar motif runs.
 *
 * `motifSize` controls the self-similar atom length. 5 is maximal for single-byte
 * motif encoding because 3^5 = 243.
 */
export function encodeSelfSimilarTritons(
  payload: readonly Triton[],
  motifSize: number = 5
): Uint8Array {
  const m = ensureMotifSize(motifSize);
  const motifs: number[] = [];
  for (let i = 0; i < payload.length; i += m) {
    motifs.push(encodeMotifByte(payload, i, m));
  }

  const body: number[] = [];
  let runCount = 0;
  if (motifs.length > 0) {
    let current = motifs[0]!;
    let length = 1;
    for (let i = 1; i < motifs.length; i++) {
      const next = motifs[i]!;
      if (next === current) {
        length += 1;
      } else {
        body.push(current);
        writeVarUint(body, length);
        runCount += 1;
        current = next;
        length = 1;
      }
    }
    body.push(current);
    writeVarUint(body, length);
    runCount += 1;
  }

  const output = new Uint8Array(SELF_SIMILAR_HEADER_SIZE + body.length);
  output.set(SELF_SIMILAR_MAGIC, 0);
  output[4] = SELF_SIMILAR_VERSION;
  output[5] = m;
  writeU32BE(output, 6, payload.length);
  writeU32BE(output, 10, runCount);
  output.set(body, SELF_SIMILAR_HEADER_SIZE);
  return output;
}

/**
 * Handles the gnosis decode Self Similar Tritons workflow.
 */
export function decodeSelfSimilarTritons(payload: Uint8Array): Triton[] {
  if (payload.length < SELF_SIMILAR_HEADER_SIZE) {
    throw new RangeError('payload too short for self-similar triton header');
  }
  for (let i = 0; i < SELF_SIMILAR_MAGIC.length; i++) {
    if (payload[i] !== SELF_SIMILAR_MAGIC[i]) {
      throw new RangeError('invalid self-similar triton magic');
    }
  }
  if (payload[4] !== SELF_SIMILAR_VERSION) {
    throw new RangeError(
      `unsupported self-similar triton version ${payload[4]}`
    );
  }
  const motifSize = ensureMotifSize(payload[5]!);
  const tritonLength = readU32BE(payload, 6);
  const runCount = readU32BE(payload, 10);
  const expectedMotifs = tritonLength === 0 ? 0 : Math.ceil(tritonLength / motifSize);
  const maxMotifValue = pow3(motifSize);

  const output: Triton[] = [];
  let motifsDecoded = 0;
  let offset = SELF_SIMILAR_HEADER_SIZE;

  for (let run = 0; run < runCount; run++) {
    if (offset >= payload.length) {
      throw new RangeError('unexpected end of payload while reading motif byte');
    }
    const motifValue = payload[offset]!;
    offset += 1;
    if (motifValue >= maxMotifValue) {
      throw new RangeError(`invalid motif byte ${motifValue} for motifSize=${motifSize}`);
    }
    const { value: runLength, nextOffset } = readVarUint(payload, offset);
    offset = nextOffset;
    if (runLength <= 0) {
      throw new RangeError('runLength must be strictly positive');
    }

    motifsDecoded += runLength;
    if (motifsDecoded > expectedMotifs) {
      throw new RangeError('decoded motifs exceed expected payload length');
    }

    const motif = decodeMotifByte(motifValue, motifSize);
    for (let i = 0; i < runLength; i++) {
      for (const symbol of motif) {
        if (output.length < tritonLength) output.push(symbol);
      }
    }
  }

  if (motifsDecoded !== expectedMotifs) {
    throw new RangeError(
      `decoded motif count mismatch: expected ${expectedMotifs}, received ${motifsDecoded}`
    );
  }
  if (output.length !== tritonLength) {
    throw new RangeError(
      `decoded triton length mismatch: expected ${tritonLength}, received ${output.length}`
    );
  }
  if (offset !== payload.length) {
    throw new RangeError('unexpected trailing bytes in self-similar payload');
  }
  return output;
}

/**
 * Handles the gnosis compress Triton Self Similar workflow.
 */
export async function compressTritonSelfSimilar(
  payload: readonly Triton[],
  options: TritonSelfSimilarCompressionOptions = {}
): Promise<TritonSelfSimilarPacket> {
  const motifSize = options.motifSize ?? 5;
  const core = encodeSelfSimilarTritons(payload, motifSize);
  const preferSmallerPayload = options.preferSmallerPayload ?? true;
  const useBitwiseFold = options.useBitwiseFold ?? true;

  let encoding: TritonSelfSimilarEncoding = 'self-similar-rle';
  let compressedPayload = core;

  if (useBitwiseFold) {
    const folded = await tryFoldWithBitwiseBinary(core);
    if (folded && (!preferSmallerPayload || folded.length < core.length)) {
      encoding = 'self-similar-rle+bitwise-fold';
      compressedPayload = folded;
    }
  }

  const rawLength = payload.length;
  const payloadLength = compressedPayload.length;

  return {
    encoding,
    motifSize: ensureMotifSize(motifSize),
    tritonLength: payload.length,
    rawLength,
    coreLength: core.length,
    payloadLength,
    compressionRatio:
      rawLength === 0 ? 1 : rawLength / Math.max(1, payloadLength),
    coreFingerprintHex: cacheFingerprint64HexFrfSync(core),
    payloadFingerprintHex: cacheFingerprint64HexFrfSync(compressedPayload),
    payload: compressedPayload,
  };
}

/**
 * Handles the gnosis decompress Triton Self Similar workflow.
 */
export async function decompressTritonSelfSimilar(
  packet: TritonSelfSimilarPacket
): Promise<Triton[]> {
  let core = packet.payload;
  if (packet.encoding === 'self-similar-rle+bitwise-fold') {
    try {
      const bitwise = await loadBitwisePkgModule();
      core = bitwise.syzygy_unfold_binary(packet.payload, packet.coreLength);
    } catch (error) {
      throw new Error(
        `bitwise-folded packet cannot be decoded without bitwise wasm bindings`,
        { cause: error }
      );
    }
  }
  if (core.length !== packet.coreLength) {
    throw new RangeError(
      `core length mismatch: expected ${packet.coreLength}, received ${core.length}`
    );
  }
  const decoded = decodeSelfSimilarTritons(core);
  if (decoded.length !== packet.tritonLength) {
    throw new RangeError(
      `triton length mismatch: expected ${packet.tritonLength}, received ${decoded.length}`
    );
  }
  return decoded;
}

/**
 * Handles the gnosis compress Triton Self Similar Batch workflow.
 */
export async function compressTritonSelfSimilarBatch(
  payloads: readonly (readonly Triton[])[],
  options: TritonSelfSimilarCompressionOptions = {}
): Promise<TritonSelfSimilarBatch> {
  const packets: TritonSelfSimilarPacket[] = [];
  for (const payload of payloads) {
    packets.push(await compressTritonSelfSimilar(payload, options));
  }
  const totalRawLength = packets.reduce((sum, packet) => sum + packet.rawLength, 0);
  const totalCompressedLength = packets.reduce(
    (sum, packet) => sum + packet.payloadLength,
    0
  );
  const combinedPayload = concatUint8Arrays(packets.map((packet) => packet.payload));
  return {
    packets,
    totalRawLength,
    totalCompressedLength,
    compressionRatio:
      totalRawLength === 0 ? 1 : totalRawLength / Math.max(1, totalCompressedLength),
    aggregateFingerprintHex: cacheFingerprint64HexFrfSync(combinedPayload),
  };
}

/**
 * Handles the gnosis decompress Triton Self Similar Batch workflow.
 */
export async function decompressTritonSelfSimilarBatch(
  batch: TritonSelfSimilarBatch
): Promise<Triton[][]> {
  const output: Triton[][] = [];
  for (const packet of batch.packets) {
    output.push(await decompressTritonSelfSimilar(packet));
  }
  return output;
}

// ---------------------------------------------------------------------------
// Part VIII: Vacuum seed + laminar / meta-laminar flow
// ---------------------------------------------------------------------------

export const VACUUM_DEFICIT = 0;
export const VACUUM_SEED = 1;

/** Deficit mass in a triton stream: non-zero trits carry unresolved pressure. */
export function tritonDeficit(payload: readonly Triton[]): number {
  return payload.reduce<number>((sum, t) => sum + (t === 0 ? 0 : 1), 0);
}

export interface TritonTeleportChannel {
  deficit: number;
  transmittedAt: number;
}

/** Deficit-only teleportation payload (minimal sufficient statistic). */
export function tritonTeleportSend(payload: readonly Triton[]): TritonTeleportChannel {
  return {
    deficit: tritonDeficit(payload),
    transmittedAt: Date.now(),
  };
}

/** Receiver deficit at round k from deficit only. */
export function tritonReceiverDeficitAt(deficit: number, round: number): number {
  const d = Math.max(0, Math.floor(deficit));
  const r = Math.max(0, Math.floor(round));
  return Math.max(0, d - Math.min(r, d));
}

/**
 * Handles the gnosis triton Teleport Trajectory workflow.
 */
export function tritonTeleportTrajectory(
  deficit: number,
  horizon: number = Math.max(0, Math.floor(deficit))
): number[] {
  const h = Math.max(0, Math.floor(horizon));
  const out: number[] = [];
  for (let k = 0; k <= h; k++) out.push(tritonReceiverDeficitAt(deficit, k));
  return out;
}

/** Meaning potential from deficit and observation budget (seeded by +1). */
export function meaningPotential(rounds: number, deficit: number): number {
  const r = Math.max(0, Math.floor(rounds));
  const d = Math.max(0, Math.floor(deficit));
  return r - Math.min(d, r) + VACUUM_SEED;
}

/** First-order laminar extraction (resolved deficit per step). */
export function laminarFlowAt(deficit: number, round: number): number {
  return (
    tritonReceiverDeficitAt(deficit, round) -
    tritonReceiverDeficitAt(deficit, round + 1)
  );
}

/** Second-order smoothness (meta-laminarity). */
export function metaLaminarAt(deficit: number, round: number): number {
  return Math.abs(laminarFlowAt(deficit, round) - laminarFlowAt(deficit, round + 1));
}

export interface VoidPumpStep {
  round: number;
  deficitBefore: number;
  deficitAfter: number;
  laminarFlow: number;
  metaLaminar: number;
  meaningPotential: number;
}

export interface VoidPumpResult {
  initialDeficit: number;
  finalDeficit: number;
  extractedMeaning: number;
  steps: VoidPumpStep[];
  laminarBoundHolds: boolean;
  metaLaminarBoundHolds: boolean;
}

/**
 * Pump the void for information with canonical deficit-only laminar dynamics.
 * This models "order wrenched from disorder" while tracking first and second
 * order smoothness.
 */
export function pumpVoidForMeaning(
  initialDeficit: number,
  horizon: number = Math.max(0, Math.floor(initialDeficit))
): VoidPumpResult {
  const d0 = Math.max(0, Math.floor(initialDeficit));
  const h = Math.max(0, Math.floor(horizon));
  const steps: VoidPumpStep[] = [];

  for (let k = 0; k < h; k++) {
    const deficitBefore = tritonReceiverDeficitAt(d0, k);
    const deficitAfter = tritonReceiverDeficitAt(d0, k + 1);
    const flow = deficitBefore - deficitAfter;
    const meta = metaLaminarAt(d0, k);
    steps.push({
      round: k,
      deficitBefore,
      deficitAfter,
      laminarFlow: flow,
      metaLaminar: meta,
      meaningPotential: meaningPotential(k, deficitBefore),
    });
  }

  const finalDeficit = tritonReceiverDeficitAt(d0, h);
  const extractedMeaning = d0 - finalDeficit;
  const laminarBoundHolds = steps.every((s) => s.laminarFlow <= 1);
  const metaLaminarBoundHolds = steps.every((s) => s.metaLaminar <= 1);

  return {
    initialDeficit: d0,
    finalDeficit,
    extractedMeaning,
    steps,
    laminarBoundHolds,
    metaLaminarBoundHolds,
  };
}

/** Vacuum state births the seed meaning floor (+1). */
export function vacuumBirthsSeed(): number {
  return meaningPotential(0, VACUUM_DEFICIT);
}
