// @a0n/gnosis/src/untangling-strategy.ts
/**
 * Practical software-knot remediation guidance.
 *
 * The supporting theorem surface establishes direction, not exact thresholds:
 * positive removable crossings justify expansion, staged expansion beats naive
 * widening when that budget is real, and some knots are genuinely irreducible.
 * This module turns those results into stable tool-facing recommendations.
 */

export const VALID_GNOSIS_UNTANGLING_ACTIONS = [
  'hold',
  'direct-simplify',
  'expand-first',
  'staged-expand-first',
  'minimize-within-class',
] as const;

export type GnosisUntanglingAction =
  (typeof VALID_GNOSIS_UNTANGLING_ACTIONS)[number];

export interface GnosisUntanglingSignal {
  crossingNumber: number;
  beta1?: number;
  violationPaths?: number;
  defectFamily?: string | null;
}

export interface GnosisUntanglingRecommendation {
  action: GnosisUntanglingAction;
  culpritLikely: boolean;
  unknottingBudget: number;
  rationale: string;
  supportingTheorems: readonly string[];
}

function toNat(value: number | undefined | null): number {
  if (!Number.isFinite(value ?? Number.NaN)) {
    return 0;
  }
  return Math.max(0, Math.trunc(value ?? 0));
}

function isCompositePressure(
  defectFamily: string | null,
  crossingNumber: number,
  violationPaths: number
): boolean {
  if (
    defectFamily === 'composite' ||
    defectFamily === 'void' ||
    defectFamily === 'periodic'
  ) {
    return true;
  }

  return crossingNumber >= 8 || violationPaths >= 2;
}

/**
 * Formats the Untangling Action Label.
 */
export function formatUntanglingActionLabel(
  action: GnosisUntanglingAction
): string {
  switch (action) {
    case 'hold':
      return 'hold';
    case 'direct-simplify':
      return 'direct simplify';
    case 'expand-first':
      return 'expand first';
    case 'staged-expand-first':
      return 'staged expand first';
    case 'minimize-within-class':
      return 'minimize within class';
  }
}

/**
 * Handles the gnosis recommend Untangling Strategy workflow.
 */
export function recommendUntanglingStrategy(
  signal: GnosisUntanglingSignal
): GnosisUntanglingRecommendation {
  const crossingNumber = toNat(signal.crossingNumber);
  const beta1 = Math.min(crossingNumber, toNat(signal.beta1));
  const violationPaths = toNat(signal.violationPaths);
  const defectFamily = signal.defectFamily ?? null;
  const unknottingBudget = Math.max(0, crossingNumber - beta1);

  if (crossingNumber === 0) {
    return {
      action: 'hold',
      culpritLikely: false,
      unknottingBudget,
      rationale:
        'Zero crossings observed. The local surface is already at the unknot.',
      supportingTheorems: ['THM-BUG-FREE-is-UNKNOT'],
    };
  }

  if (beta1 === crossingNumber) {
    return {
      action: 'minimize-within-class',
      culpritLikely: false,
      unknottingBudget,
      rationale:
        'Every observed crossing is essential loop structure. Bound the knot and minimize it in-class instead of promising full elimination.',
      supportingTheorems: ['THM-HYPERBOLIC-HAS-MINIMUM'],
    };
  }

  if (
    unknottingBudget >= 4 &&
    isCompositePressure(defectFamily, crossingNumber, violationPaths)
  ) {
    return {
      action: 'staged-expand-first',
      culpritLikely: true,
      unknottingBudget,
      rationale:
        'High removable crossing budget plus composite pressure. Split responsibility into explicit stages or helper boundaries first, then collapse once the untie move is visible.',
      supportingTheorems: [
        'THM-S7-STAGGER',
        'THM-STAGED-EXPANSION',
        'THM-COMPOSITION-STRICTLY-MONOTONE',
      ],
    };
  }

  if (unknottingBudget >= 2) {
    return {
      action: 'expand-first',
      culpritLikely: crossingNumber >= 6,
      unknottingBudget,
      rationale:
        'Positive removable crossing budget detected. Introduce an explicit intermediate state, helper, or adapter before simplifying.',
      supportingTheorems: ['THM-STAGED-EXPANSION'],
    };
  }

  return {
    action: 'direct-simplify',
    culpritLikely: false,
    unknottingBudget,
    rationale:
      'The knot is shallow enough that a local untie move should be reachable without state-space expansion.',
    supportingTheorems: ['THM-FIX-is-REIDEMEISTER'],
  };
}
