// @a0n/gnosis/src/valentinian-sin.ts
/**
 * Valentinian Sin Diagnostic & Cross-Pollination
 * 
 * Implements an automated mathematical gauge to verify if a process graph
 * has become detached from its local environment (void-isolated).
 * Cross-Pollination safely bridges isolated components back into reality.
 */

export interface ValentinianState {
  voidTicks: number;
  isolationThreshold: number;
}

/**
 * Valentinian Sin: "The error of the Pleroma was detachment."
 * Returns true if an agent operates purely on internal state for too long.
 */
export function assertValentinianSin(state: ValentinianState): boolean {
    return state.voidTicks >= state.isolationThreshold;
}

/**
 * Measures the "Truth Debt" accumulated by ignoring the sensorium.
 * Exponential penalty as isolation grows.
 */
export function computeIgnorancePenalty(state: ValentinianState): number {
    if (state.voidTicks === 0) return 0;
    return Math.pow(2, state.voidTicks / state.isolationThreshold);
}

/**
 * Valentinian Cross-Pollination
 * 
 * When an agent reaches the Sin boundary, it MUST execute a forced FOLD or VENT
 * to cross-pollinate with the external reality, resetting the void ticks.
 */
export function performCrossPollination(state: ValentinianState): ValentinianState {
    // A structural re-basing against the external reality sets divergence back to 0
    return {
        ...state,
        voidTicks: 0,
    };
}
