// @a0n/gnosis/src/void-equilibrium-geometry.ts
/**
 * The Void — Equilibrium Geometry & Void Estate Economy
 *
 * Implements rigorous mathematical spatial scattering of Nash and Skyrms
 * geographic equilibriums across the Cartesian Hex Grid. Uses deterministic
 * hashing bounded by Prime constants to guarantee stability without manual
 * database tracking.
 *
 * @module @a0n/gnosis/void-equilibrium-geometry
 */

import crypto from 'crypto';
import type { GameCoordinate } from './void-game-types';

export type EquilibriumType = 'none' | 'nash' | 'skyrms' | 'blackhole';

const VOID_GENESIS_DID = 'did:key:void-genesis';

/**
 * Derives a deterministic mathematical hash using the absolute spatial
 * dimension and geometry. The hash acts as a geographical RNG that never changes.
 */
function HashCoordinateGeometry(coordinate: GameCoordinate): number {
    const geoStr = `dim:${coordinate.dimension}:q=${coordinate.hexQ}:r=${coordinate.hexR}:s=${coordinate.hexS}`;
    const hashHex = crypto.createHash('md5').update(geoStr).digest('hex').substring(0, 8);
    return parseInt(hashHex, 16);
}

function sphericalToHexCubeStr(raDeg: number, decDeg: number, dist: number): string {
    const raRad = (raDeg * Math.PI) / 180;
    const decRad = (decDeg * Math.PI) / 180;
    
    // Scale down lightyears/AU into playable game grid ticks
    const r = Math.max(1, Math.round(dist));
    
    const x = Math.round(r * Math.cos(decRad) * Math.cos(raRad));
    const y = Math.round(r * Math.cos(decRad) * Math.sin(raRad));
    // Enforce q + r + s = 0 constraint for Hex Grid
    const s = -x - y;
    
    return `${x},${y},${s}`;
}

// Ported from aeon-solarium Stars-Fish Atlas 
const BLACK_HOLE_SNAPSHOT = [
  { id: 'a0620-00', ra: 95.685, dec: -0.346, dist: 3300 },
  { id: 'cygnus-x-1', ra: 299.59, dec: 35.202, dist: 7200 },
  { id: 'sagittarius-a-star', ra: 266.4168, dec: -29.0078, dist: 26673 },
  { id: 'm87-star', ra: 187.7059, dec: 12.3911, dist: 53000000 },
];

const SOLAR_SYSTEM_SNAPSHOT = [
  { id: 'mercury', ra: 252.25084, dec: 7.00487, dist: 387 }, 
  { id: 'venus', ra: 181.97973, dec: 3.39471, dist: 723 },
  { id: 'earth', ra: 100.46435, dec: 0.00005, dist: 1000 },
  { id: 'mars', ra: -4.553432, dec: 1.84969, dist: 1523 },
  { id: 'jupiter', ra: 34.40438, dec: 1.30327, dist: 5202 },
  { id: 'saturn', ra: 49.94432, dec: 2.48888, dist: 9554 },
  { id: 'uranus', ra: 313.23218, dec: 0.7732, dist: 19218 },
  { id: 'neptune', ra: -55.12003, dec: 1.76917, dist: 30110 }
];

const ATLAS_BLACK_HOLES = new Set<string>();
const ATLAS_PLANETS = new Set<string>();

// Precompute Exact HexCoordinate Mappings At Runtime
for (const bh of BLACK_HOLE_SNAPSHOT) {
    ATLAS_BLACK_HOLES.add(sphericalToHexCubeStr(bh.ra, bh.dec, bh.dist));
}
for (const p of SOLAR_SYSTEM_SNAPSHOT) {
    ATLAS_PLANETS.add(sphericalToHexCubeStr(p.ra, p.dec, p.dist));
}

/**
 * Mathematically proves if the passed physical coordinate overlaps
 * a massive Game Theory topographical anomaly.
 * Now driven by the Stars-Fish Universal Atlas rules!
 */
export function identifyEquilibriumGeometry(coordinate: GameCoordinate): EquilibriumType {
    // Only map exact coordinates on Dimension 3 (Physical Canonical Layer) 
    if (coordinate.dimension === 3) {
        const key = `${coordinate.hexQ},${coordinate.hexR},${coordinate.hexS}`;
        
        if (ATLAS_BLACK_HOLES.has(key)) {
            return 'blackhole';
        }
        if (ATLAS_PLANETS.has(key)) {
            return 'skyrms';
        }
    }

    const geoHash = HashCoordinateGeometry(coordinate);

    // Fall back to Nash equilibria generation using the old secure geographic RNG
    if (geoHash % 1009 === 0) {
        return 'nash';
    }

    return 'none';
}

/**
 * Checks if a player/agent illegally attempts to initialize an infrastructure
 * node on Prime Real-Estate that natively belongs to the Void.
 * 
 * If the estate is owned by void-genesis, the buyer MUST utilize the 
 * gnosis_transfer_infrastructure MCP tool to pay the thermodynamic cost 
 * of transferring ownership. Normal deployment is blocked.
 */
export function validateGenesisOwnershipBoundaries(coordinate: GameCoordinate, claimantDid: string): { valid: boolean, requiredOwner?: string } {
    const equilibrium = identifyEquilibriumGeometry(coordinate);
    
    // If it's a completely standard mathematical coordinate, any agent can deploy.
    if (equilibrium === 'none') {
        return { valid: true };
    }

    // It is an Epic/Legendary location. At Universe Genesis, the Void owns it.
    // Assuming the user has not physically swapped titles yet (checked at a higher DASH state layer).
    // The physics engine fundamentally believes it belongs to Genesis.
    return { 
        valid: claimantDid === VOID_GENESIS_DID,
        requiredOwner: VOID_GENESIS_DID 
    };
}
