// @a0n/gnosis/src/void-attention-handlers.ts
/**
 * void-attention-handlers.ts -- Handlers that make void-attention.gg executable
 *
 * When registered with the GnosisRegistry, these handlers allow the gnosis
 * engine to execute the void-attention topology. Each handler implements
 * one node type from the identification:
 *
 *   complement_distribution  →  softmax attention score
 *   outer_product            →  cross-attention join
 *   gated_product            →  gated cross-attention (Skyrms gate)
 *   spread_rejection         →  attention pattern neighborhood
 *   void_decay               →  layer norm
 *   measure                  →  c1 (kurtosis, entropy, inverse Bule)
 *   evaluate_payoff           →  interaction evaluation
 *   accept_if_higher_weight  →  decision (accept proposal vs own choice)
 *   argmax                   →  greedy decode from complement
 */

import type { GnosisRegistry, GnosisHandler } from './runtime/registry';
import {
  type VoidBoundary,
  type Walker,
  type Measurement,
  type Gait,
  createVoidBoundary,
  createWalker,
  updateVoidBoundary,
  decayVoidBoundary,
  complementDistribution,
  sampleComplement,
  shannonEntropy,
  excessKurtosis,
  giniCoefficient,
  inverseBule,
  measure,
  c0Update,
  c1Measure,
  c2c3Adapt,
} from './void';

// ============================================================================
// Payload shape flowing through the void-attention topology
// ============================================================================

export interface VoidAttentionPayload {
  /** Walker A's void boundary */
  boundaryA: VoidBoundary;
  /** Walker B's void boundary */
  boundaryB: VoidBoundary;
  /** Skyrms walker's void boundary (over proposal space) */
  boundaryCross: VoidBoundary;
  /** Walkers */
  walkerA: Walker;
  walkerB: Walker;
  walkerCross: Walker;
  /** Dimensions */
  dimA: number;
  dimB: number;
  /** Complement distributions (computed during attention) */
  complementA?: number[];
  complementB?: number[];
  complementCross?: number[];
  /** Joint surface (outer product * gate) */
  jointSurface?: number[];
  /** Selected offers */
  proposalA?: number;
  proposalB?: number;
  ownChoiceA?: number;
  ownChoiceB?: number;
  offerA?: number;
  offerB?: number;
  /** Proposal acceptance */
  proposalAccepted?: boolean;
  /** Payoffs from interaction */
  payoffA?: number;
  payoffB?: number;
  /** Neighborhood radius (scales with gait) */
  neighborhoodRadius: number;
  /** Decay rate */
  decayRate: number;
  /** Payoff function */
  payoff: (a: number, b: number) => [number, number];
  /** RNG */
  rng: () => number;
  /** Round counter */
  round: number;
}

// ============================================================================
// Self-Attention Handlers
// ============================================================================

/** Compute complement distribution = softmax(-eta * voidCounts) */
const ComplementDistributionAHandler: GnosisHandler = async (
  p: VoidAttentionPayload
) => {
  const dist = complementDistribution(p.boundaryA, p.walkerA.eta);
  return { ...p, complementA: dist };
};

const ComplementDistributionBHandler: GnosisHandler = async (
  p: VoidAttentionPayload
) => {
  const dist = complementDistribution(p.boundaryB, p.walkerB.eta);
  return { ...p, complementB: dist };
};

const ComplementDistributionCrossHandler: GnosisHandler = async (
  p: VoidAttentionPayload
) => {
  const dist = complementDistribution(p.boundaryCross, p.walkerCross.eta);
  return { ...p, complementCross: dist };
};

// ============================================================================
// Cross-Attention Handlers
// ============================================================================

/** Outer product of two marginal complement distributions */
const OuterProductHandler: GnosisHandler = async (p: VoidAttentionPayload) => {
  if (!p.complementA || !p.complementB) return p;
  const surface = new Array(p.dimA * p.dimB);
  for (let i = 0; i < p.dimA; i++) {
    for (let j = 0; j < p.dimB; j++) {
      surface[i * p.dimB + j] = p.complementA[i] * p.complementB[j];
    }
  }
  return { ...p, jointSurface: surface };
};

/** Gated cross-attention: joint surface * Skyrms gate, normalized */
const GatedProductHandler: GnosisHandler = async (p: VoidAttentionPayload) => {
  if (!p.jointSurface || !p.complementCross) return p;
  const N = p.jointSurface.length;
  const gated = new Array(N);
  let sum = 0;
  for (let k = 0; k < N; k++) {
    gated[k] = p.jointSurface[k] * p.complementCross[k];
    sum += gated[k];
  }
  if (sum > 0) for (let k = 0; k < N; k++) gated[k] /= sum;
  return { ...p, jointSurface: gated };
};

// ============================================================================
// Decision Handlers
// ============================================================================

/** Sample from complement distribution (with exploration) */
const SampleAHandler: GnosisHandler = async (p: VoidAttentionPayload) => {
  if (!p.complementA) return p;
  // Epsilon-greedy: explore or exploit
  if (p.rng() < p.walkerA.exploration) {
    return { ...p, ownChoiceA: Math.floor(p.rng() * p.dimA) };
  }
  return { ...p, ownChoiceA: sampleComplement(p.complementA, p.rng) };
};

const SampleBHandler: GnosisHandler = async (p: VoidAttentionPayload) => {
  if (!p.complementB) return p;
  if (p.rng() < p.walkerB.exploration) {
    return { ...p, ownChoiceB: Math.floor(p.rng() * p.dimB) };
  }
  return { ...p, ownChoiceB: sampleComplement(p.complementB, p.rng) };
};

/** Sample from joint surface → proposal */
const SampleCrossHandler: GnosisHandler = async (p: VoidAttentionPayload) => {
  if (!p.jointSurface) return p;
  if (p.rng() < p.walkerCross.exploration) {
    const flat = Math.floor(p.rng() * p.jointSurface.length);
    return {
      ...p,
      proposalA: Math.floor(flat / p.dimB),
      proposalB: flat % p.dimB,
    };
  }
  const flat = sampleComplement(p.jointSurface, p.rng);
  return {
    ...p,
    proposalA: Math.floor(flat / p.dimB),
    proposalB: flat % p.dimB,
  };
};

/** Accept proposal if complement weight >= own choice weight */
const AcceptIfHigherWeightHandler: GnosisHandler = async (
  p: VoidAttentionPayload
) => {
  if (!p.complementA || !p.complementB) return p;
  if (p.proposalA === undefined || p.proposalB === undefined) return p;
  if (p.ownChoiceA === undefined || p.ownChoiceB === undefined) return p;

  const offerA =
    p.complementA[p.proposalA] >= p.complementA[p.ownChoiceA]
      ? p.proposalA
      : p.ownChoiceA;
  const offerB =
    p.complementB[p.proposalB] >= p.complementB[p.ownChoiceB]
      ? p.proposalB
      : p.ownChoiceB;

  return {
    ...p,
    offerA,
    offerB,
    proposalAccepted: offerA === p.proposalA && offerB === p.proposalB,
  };
};

// ============================================================================
// Interaction Handler
// ============================================================================

const EvaluatePayoffHandler: GnosisHandler = async (
  p: VoidAttentionPayload
) => {
  if (p.offerA === undefined || p.offerB === undefined) return p;
  const [payA, payB] = p.payoff(p.offerA, p.offerB);
  return { ...p, payoffA: payA, payoffB: payB };
};

// ============================================================================
// Residual Update Handlers (void boundary absorbs outcome)
// ============================================================================

const ResidualUpdateAHandler: GnosisHandler = async (
  p: VoidAttentionPayload
) => {
  if (
    p.offerA === undefined ||
    p.payoffA === undefined ||
    p.payoffB === undefined
  )
    return p;
  if (p.payoffA < p.payoffB || p.payoffA < 0) {
    const mag = p.payoffA < 0 ? Math.abs(p.payoffA) : 1;
    updateVoidBoundary(p.boundaryA, p.offerA, mag);
    p.walkerA.steps++;
  }
  // Cross-pollinate
  if (p.offerA !== p.offerB && p.offerB !== undefined) {
    updateVoidBoundary(p.boundaryA, Math.min(p.offerB, p.dimA - 1), 1);
  }
  p.walkerA.score += p.payoffA;
  return p;
};

const ResidualUpdateBHandler: GnosisHandler = async (
  p: VoidAttentionPayload
) => {
  if (
    p.offerB === undefined ||
    p.payoffA === undefined ||
    p.payoffB === undefined
  )
    return p;
  if (p.payoffB < p.payoffA || p.payoffB < 0) {
    const mag = p.payoffB < 0 ? Math.abs(p.payoffB) : 1;
    updateVoidBoundary(p.boundaryB, p.offerB, mag);
    p.walkerB.steps++;
  }
  if (p.offerA !== p.offerB && p.offerA !== undefined) {
    updateVoidBoundary(p.boundaryB, Math.min(p.offerA, p.dimB - 1), 1);
  }
  p.walkerB.score += p.payoffB;
  return p;
};

const ResidualUpdateCrossHandler: GnosisHandler = async (
  p: VoidAttentionPayload
) => {
  if (p.proposalA === undefined || p.proposalB === undefined) return p;
  if (!p.proposalAccepted || p.offerA !== p.offerB) {
    const flat = p.proposalA * p.dimB + p.proposalB;
    const magnitude = p.proposalAccepted ? 1 : 2;
    updateVoidBoundary(p.boundaryCross, flat, magnitude);
    p.walkerCross.steps++;

    // Neighborhood poisoning
    const r = p.neighborhoodRadius;
    if (r > 0) {
      for (let da = -r; da <= r; da++) {
        for (let db = -r; db <= r; db++) {
          if (da === 0 && db === 0) continue;
          const nA = p.proposalA + da;
          const nB = p.proposalB + db;
          if (nA >= 0 && nA < p.dimA && nB >= 0 && nB < p.dimB) {
            const neighborFlat = nA * p.dimB + nB;
            const dist = Math.abs(da) + Math.abs(db);
            c0Update(
              p.walkerCross,
              neighborFlat,
              Math.max(1, Math.round(magnitude / dist))
            );
          }
        }
      }
    }
  }
  return p;
};

// ============================================================================
// Layer Norm Handler (void decay)
// ============================================================================

const VoidDecayHandler: GnosisHandler = async (p: VoidAttentionPayload) => {
  if (p.decayRate > 0) {
    decayVoidBoundary(p.boundaryA, p.decayRate);
    decayVoidBoundary(p.boundaryB, p.decayRate);
    decayVoidBoundary(p.boundaryCross, p.decayRate);
  }
  return p;
};

// ============================================================================
// Feed-Forward Handler (c3 gait adaptation)
// ============================================================================

const GaitAdaptHandler: GnosisHandler = async (p: VoidAttentionPayload) => {
  c2c3Adapt(p.walkerA);
  c2c3Adapt(p.walkerB);
  c2c3Adapt(p.walkerCross);
  p.round++;
  return p;
};

async function invokeVoidAttentionHandler(
  handler: GnosisHandler,
  payload: VoidAttentionPayload
): Promise<VoidAttentionPayload> {
  return handler(payload, {}, undefined) as Promise<VoidAttentionPayload>;
}

// ============================================================================
// Measurement Handler (c1)
// ============================================================================

const MeasureHandler: GnosisHandler = async (p: VoidAttentionPayload) => {
  return {
    a: measure(p.boundaryA, p.walkerA.eta, p.walkerA.steps),
    b: measure(p.boundaryB, p.walkerB.eta, p.walkerB.steps),
    cross: measure(p.boundaryCross, p.walkerCross.eta, p.walkerCross.steps),
  };
};

// ============================================================================
// Registration
// ============================================================================

/**
 * Register all void-attention handlers with a GnosisRegistry.
 * After registration, the void-attention.gg topology can be compiled
 * and executed by the gnosis engine.
 */
export function registerVoidAttentionHandlers(registry: GnosisRegistry): void {
  // Self-attention
  registry.register('ComplementDistributionA', ComplementDistributionAHandler);
  registry.register('ComplementDistributionB', ComplementDistributionBHandler);
  registry.register(
    'ComplementDistributionCross',
    ComplementDistributionCrossHandler
  );

  // Cross-attention
  registry.register('OuterProduct', OuterProductHandler);
  registry.register('GatedProduct', GatedProductHandler);

  // Decision
  registry.register('SampleA', SampleAHandler);
  registry.register('SampleB', SampleBHandler);
  registry.register('SampleCross', SampleCrossHandler);
  registry.register('AcceptIfHigherWeight', AcceptIfHigherWeightHandler);

  // Interaction
  registry.register('EvaluatePayoff', EvaluatePayoffHandler);

  // Residual (void boundary update)
  registry.register('ResidualUpdateA', ResidualUpdateAHandler);
  registry.register('ResidualUpdateB', ResidualUpdateBHandler);
  registry.register('ResidualUpdateCross', ResidualUpdateCrossHandler);

  // Layer norm
  registry.register('VoidDecay', VoidDecayHandler);

  // Feed-forward
  registry.register('GaitAdapt', GaitAdaptHandler);

  // Measurement
  registry.register('VoidMeasure', MeasureHandler);
}

/**
 * Create an initial payload for executing void-attention.gg.
 */
export function createVoidAttentionPayload(
  dimA: number,
  dimB: number,
  payoff: (a: number, b: number) => [number, number],
  rng: () => number = Math.random,
  neighborhoodRadius = 1,
  decayRate = 0
): VoidAttentionPayload {
  const boundaryA = createVoidBoundary(dimA);
  const boundaryB = createVoidBoundary(dimB);
  const boundaryCross = createVoidBoundary(dimA * dimB);
  return {
    boundaryA,
    boundaryB,
    boundaryCross,
    walkerA: createWalker(boundaryA),
    walkerB: createWalker(boundaryB),
    walkerCross: createWalker(boundaryCross),
    dimA,
    dimB,
    neighborhoodRadius,
    decayRate,
    payoff,
    rng,
    round: 0,
  };
}

/**
 * Run one full forward pass of the void attention transformer.
 * Executes all handlers in sequence (equivalent to engine execution
 * of the void-attention.gg topology).
 */
export async function voidAttentionForward(
  payload: VoidAttentionPayload
): Promise<VoidAttentionPayload> {
  let p = payload;

  // 1. Self-attention
  p = await invokeVoidAttentionHandler(ComplementDistributionAHandler, p);
  p = await invokeVoidAttentionHandler(ComplementDistributionBHandler, p);
  p = await invokeVoidAttentionHandler(ComplementDistributionCrossHandler, p);

  // 2. Cross-attention
  p = await invokeVoidAttentionHandler(OuterProductHandler, p);
  p = await invokeVoidAttentionHandler(GatedProductHandler, p);

  // 3. Decision
  p = await invokeVoidAttentionHandler(SampleAHandler, p);
  p = await invokeVoidAttentionHandler(SampleBHandler, p);
  p = await invokeVoidAttentionHandler(SampleCrossHandler, p);
  p = await invokeVoidAttentionHandler(AcceptIfHigherWeightHandler, p);

  // 4. Interaction
  p = await invokeVoidAttentionHandler(EvaluatePayoffHandler, p);

  // 5. Residual update
  p = await invokeVoidAttentionHandler(ResidualUpdateAHandler, p);
  p = await invokeVoidAttentionHandler(ResidualUpdateBHandler, p);
  p = await invokeVoidAttentionHandler(ResidualUpdateCrossHandler, p);

  // 6. Layer norm
  p = await invokeVoidAttentionHandler(VoidDecayHandler, p);

  // 7. Feed-forward
  p = await invokeVoidAttentionHandler(GaitAdaptHandler, p);

  return p;
}
