// @a0n/gnosis/src/void-handlers.ts
/**
 * void-handlers.ts -- Handlers that make void-walking.gg executable
 *
 * When registered with the GnosisRegistry, these handlers allow
 * the gnosis engine to execute the self-referential void-walking
 * topology. The topology describes the algorithm; these handlers
 * implement each node's computation.
 *
 * The result: void walking, described as a .gg topology, compiled
 * by Betty, executed by the gnosis engine, using void walking to
 * decide execution order. Gnosis eats itself.
 */

import type { GnosisRegistry, GnosisHandler } from './runtime/registry';
import {
  type VoidBoundary,
  type Walker,
  type Measurement,
  type Gait,
  createVoidBoundary,
  createWalker,
  updateVoidBoundary,
  complementDistribution,
  sampleComplement,
  shannonEntropy,
  excessKurtosis,
  giniCoefficient,
  inverseBule,
//   selectGait,
//   measure,
//   GAIT_DEPTH,
} from './void';

// ============================================================================
// Payload shape flowing through the topology
// ============================================================================

export interface VoidWalkingPayload {
  boundary: VoidBoundary;
  walker: Walker;
  rng: () => number;
  /** Last computed complement distribution */
  complement?: number[];
  /** Last chosen dimension index */
  chosenIdx?: number;
  /** Last measurement */
  measurement?: Measurement;
  /** Current gait */
  gait?: Gait;
  /** Iteration count */
  iteration: number;
}

// ============================================================================
// Handler implementations
// ============================================================================

const VoidBoundaryHandler: GnosisHandler = async (
  payload: VoidWalkingPayload
) => {
  // Entry point: ensure the boundary exists
  return {
    ...payload,
    iteration: (payload.iteration ?? 0) + 1,
  };
};

const ComplementDistributionHandler: GnosisHandler = async (
  payload: VoidWalkingPayload
) => {
  const dist = complementDistribution(payload.boundary, payload.walker.eta);
  return { ...payload, complement: dist };
};

const DimensionHandler: GnosisHandler = async (payload: VoidWalkingPayload) => {
  // Each Dimension node is a possible choice. During RACE, they compete.
  // The one chosen by the complement distribution wins.
  if (!payload.complement) return payload;
  const idx = sampleComplement(payload.complement, payload.rng);
  return { ...payload, chosenIdx: idx };
};

const BoundaryUpdateHandler: GnosisHandler = async (
  payload: VoidWalkingPayload
) => {
  if (payload.chosenIdx === undefined) return payload;
  // The RACE winner accumulates void at the chosen dimension
  updateVoidBoundary(payload.boundary, payload.chosenIdx);
  payload.walker.steps++;
  return payload;
};

const MeasurementHandler: GnosisHandler = async (
  payload: VoidWalkingPayload
) => {
  return payload; // Measurement node just passes through to FORK
};

const ShannonEntropyHandler: GnosisHandler = async (
  payload: VoidWalkingPayload
) => {
  if (!payload.complement) return 0;
  return shannonEntropy(payload.complement);
};

const ExcessKurtosisHandler: GnosisHandler = async (
  payload: VoidWalkingPayload
) => {
  if (!payload.complement) return 0;
  return excessKurtosis(payload.complement);
};

const GiniCoefficientHandler: GnosisHandler = async (
  payload: VoidWalkingPayload
) => {
  if (!payload.complement) return 0;
  return giniCoefficient(payload.complement);
};

const InverseBuleHandler: GnosisHandler = async (
  payload: VoidWalkingPayload
) => {
  if (!payload.boundary) return 0;
  return inverseBule(
    payload.boundary,
    payload.walker?.eta ?? 2.0,
    payload.walker?.steps ?? 0
  );
};

const ObservationHandler: GnosisHandler = async (payload: unknown) => {
  // FOLD merges the four measurement results into a Measurement object
  if (typeof payload === 'object' && payload !== null) {
    const p = payload as Record<string, unknown>;
    // The fold produces an object keyed by node ID
    const measurement: Measurement = {
      entropy: typeof p.entropy === 'number' ? p.entropy : 0,
      kurtosis: typeof p.kurtosis === 'number' ? p.kurtosis : 0,
      gini: typeof p.gini === 'number' ? p.gini : 0,
      inverseBule: typeof p.inv_bule === 'number' ? p.inv_bule : 0,
    };
    return measurement;
  }
  return payload;
};

const GaitSelectorHandler: GnosisHandler = async (payload: unknown) => {
  return payload; // Pass measurement through to SUPERPOSE
};

/** Gait handlers: each represents a possible gait state */
const StandHandler: GnosisHandler = async () => 'stand' as Gait;
const TrotHandler: GnosisHandler = async () => 'trot' as Gait;
const CanterHandler: GnosisHandler = async () => 'canter' as Gait;
const GallopHandler: GnosisHandler = async () => 'gallop' as Gait;

const ActiveGaitHandler: GnosisHandler = async (payload: unknown) => {
  // RACE winner is the selected gait
  return payload;
};

const ParameterAdaptationHandler: GnosisHandler = async (
  payload: VoidWalkingPayload
) => {
  return payload; // Pass through to FORK for eta/exploration adjustment
};

const EtaAdjustmentHandler: GnosisHandler = async (
  payload: VoidWalkingPayload
) => {
  const gait = payload.gait ?? payload.walker.gait;
  switch (gait) {
    case 'stand':
      break;
    case 'trot':
      payload.walker.eta = Math.max(1.0, payload.walker.eta - 0.05);
      break;
    case 'canter':
      payload.walker.eta = Math.min(5.0, payload.walker.eta + 0.05);
      break;
    case 'gallop':
      payload.walker.eta = Math.min(8.0, payload.walker.eta + 0.1);
      break;
  }
  return payload;
};

const ExplorationAdjustmentHandler: GnosisHandler = async (
  payload: VoidWalkingPayload
) => {
  const gait = payload.gait ?? payload.walker.gait;
  switch (gait) {
    case 'stand':
      break;
    case 'trot':
      payload.walker.exploration = Math.min(
        0.4,
        payload.walker.exploration + 0.01
      );
      break;
    case 'canter':
      payload.walker.exploration = Math.max(
        0.05,
        payload.walker.exploration - 0.005
      );
      break;
    case 'gallop':
      payload.walker.exploration = Math.max(
        0.01,
        payload.walker.exploration - 0.01
      );
      break;
  }
  return payload;
};

const AdaptedStateHandler: GnosisHandler = async (
  payload: VoidWalkingPayload
) => {
  // FOLD merges eta and exploration adjustments
  // The adapted state feeds back into the boundary (the loop)
  payload.walker.adaptations++;
  payload.walker.lastAdaptationStep = payload.walker.steps;
  return payload;
};

const ConvergedHandler: GnosisHandler = async (payload: unknown) => {
  return { converged: true, payload };
};

const DivergedHandler: GnosisHandler = async (payload: unknown) => {
  return { diverged: true, payload };
};

// ============================================================================
// Registration
// ============================================================================

/**
 * Register all void-walking handlers with a GnosisRegistry.
 * After registration, the void-walking.gg topology can be compiled
 * and executed by the gnosis engine.
 */
export function registerVoidWalkingHandlers(registry: GnosisRegistry): void {
  registry.register('VoidBoundary', VoidBoundaryHandler);
  registry.register('ComplementDistribution', ComplementDistributionHandler);
  registry.register('Dimension', DimensionHandler);
  registry.register('BoundaryUpdate', BoundaryUpdateHandler);
  registry.register('Measurement', MeasurementHandler);
  registry.register('ShannonEntropy', ShannonEntropyHandler);
  registry.register('ExcessKurtosis', ExcessKurtosisHandler);
  registry.register('GiniCoefficient', GiniCoefficientHandler);
  registry.register('InverseBule', InverseBuleHandler);
  registry.register('Observation', ObservationHandler);
  registry.register('GaitSelector', GaitSelectorHandler);
  registry.register('Stand', StandHandler);
  registry.register('Trot', TrotHandler);
  registry.register('Canter', CanterHandler);
  registry.register('Gallop', GallopHandler);
  registry.register('ActiveGait', ActiveGaitHandler);
  registry.register('ParameterAdaptation', ParameterAdaptationHandler);
  registry.register('EtaAdjustment', EtaAdjustmentHandler);
  registry.register('ExplorationAdjustment', ExplorationAdjustmentHandler);
  registry.register('AdaptedState', AdaptedStateHandler);
  registry.register('Converged', ConvergedHandler);
  registry.register('Diverged', DivergedHandler);
}

/**
 * Create an initial payload for executing void-walking.gg.
 */
export function createVoidWalkingPayload(
  dimensions: number,
  rng: () => number = Math.random
): VoidWalkingPayload {
  const boundary = createVoidBoundary(dimensions);
  return {
    boundary,
    walker: createWalker(boundary),
    rng,
    iteration: 0,
  };
}
