// @a0n/gnosis/src/void-edgework-economy.ts
/**
 * The Void — EDGEWORK Mesh Economy
 *
 * "The system pays the game for its mesh contributions, and the game uses
 * that HEX to pay its players. Always positive, but fractional..."
 *
 * The classic MMO "gold farming loop" is replaced by actual, functional
 * computational utility. The Forkjoin Monorepo runs real infrastructure 
 * via the `EDGEWORK` framework (Cloudflare/Vercel workers, UCAN syncs).
 *
 * As the real-world mesh network operates, it mathematically yields raw
 * thermodynamic `HEX`. This module computes those network contributions
 * and distributes the HEX unconditionally to active players as positive-sum yields.
 * 
 * @module @a0n/gnosis/void-edgework-economy
 */

import type { ObserverState } from './void-game-types';
import { calculateEnergyCapacity } from './void-dimensional-mastery';

export interface EdgeworkMeshTelemetry {
  readonly activeWorkerNodes: number;
  readonly packetsRelayed: number;
  readonly crdtMergesCompleted: number;
  readonly ucanVerifications: number;
}

/**
 * Calculate the total HEX generated globally by the real-worldEDGEWORK mesh.
 * This turns web infrastructure operations into pure thermodynamic energy.
 */
export function calculateGlobalNetworkYield(telemetry: EdgeworkMeshTelemetry): number {
  // Compute weightings based on actual network utility
  const nodeYield = telemetry.activeWorkerNodes * 10;
  const packetYield = telemetry.packetsRelayed * 0.1;
  const mergeYield = telemetry.crdtMergesCompleted * 2;
  const verificationYield = telemetry.ucanVerifications * 5;

  return Math.floor(nodeYield + packetYield + mergeYield + verificationYield);
}

export interface EconomyDistributionResult {
  readonly updatedObservers: ObserverState[];
  readonly yieldPerObserver: number;
}

/**
 * Distribute the global EDGEWORK computing yield completely equitably among
 * all structurally valid (online) observers.
 * 
 * "Always positive, but fractional depending on EDGEWORK contributions."
 * Observers receive HEX organically just by existing in the connected mesh,
 * effectively offsetting the natural Stagnation decay of the universe!
 */
export function distributeEdgeworkYield(
  telemetry: EdgeworkMeshTelemetry,
  activeObservers: ObserverState[]
): EconomyDistributionResult {
  if (activeObservers.length === 0) {
    return { updatedObservers: [], yieldPerObserver: 0 };
  }

  const globalYield = calculateGlobalNetworkYield(telemetry);
  
  // Universal basic thermodynamic income! 
  // Distributed fractionally across the community.
  const yieldPerObserver = Math.floor(globalYield / activeObservers.length);

  // Distribute it (bounding to their natural dimension capacity)
  const updatedObservers = activeObservers.map(obs => {
    const maxEnergy = calculateEnergyCapacity(obs.highestDimension);
    return {
      ...obs,
      energy: Math.min(maxEnergy, obs.energy + yieldPerObserver)
    };
  });

  return {
    updatedObservers,
    yieldPerObserver
  };
}
