// @a0n/gnosis/src/void-regional-shards.ts
/**
 * The Void — Regional Shard Topology
 *
 * Implements mathematical sharding of the infinite 3D hex grid into bounded regional play areas.
 * Real-time presence and physics are only computed for entities within the same (or adjacent) regional shards.
 * Asynchronous state reconciliation (dash/CRDTs) manages inter-shard continuity.
 */

import type { DimensionIndex, GameCoordinate } from './void-game-types';

/** The size of a regional shard (diameter in hexes) */
export const SHARD_SIZE = 1000;

export interface RegionalShard {
  readonly shardId: string;
  readonly dimension: DimensionIndex;
  readonly chunkQ: number;
  readonly chunkR: number;
  readonly chunkS: number;
}

/**
 * Determine which regional shard a purely theoretical GameCoordinate resides in.
 * Uses integer division to chunk the Q, R, S coordinates.
 */
export function coordinateToShard(coordinate: GameCoordinate): RegionalShard {
  // We use mathematical floor division to chunk the space into massive triangles/hexes
  // For axial coordinates, we can chunk based on Q and R since S is derived.
  // Using Math.floor(coord / SHARD_SIZE) groups SHARD_SIZE x SHARD_SIZE squares, 
  // which distorts the hex grid slightly at boundaries, but is highly deterministic and fast for routing.
  
  const chunkQ = Math.floor(coordinate.hexQ / SHARD_SIZE);
  const chunkR = Math.floor(coordinate.hexR / SHARD_SIZE);
  const chunkS = -chunkQ - chunkR; // Rederive structural zero-sum
  
  const shardId = `shard-dim${coordinate.dimension}-q${chunkQ}r${chunkR}`;
  
  return {
    shardId,
    dimension: coordinate.dimension,
    chunkQ,
    chunkR,
    chunkS,
  };
}

/**
 * Get adjacent shard IDs to subscribe to edge-case physical events.
 */
export function getAdjacentShards(shard: RegionalShard): string[] {
  const directions = [
    [1, 0], [1, -1], [0, -1], [-1, 0], [-1, 1], [0, 1]
  ];
  
  return directions.map(([dq, dr]) => {
    const q = shard.chunkQ + dq;
    const r = shard.chunkR + (dr as number);
    return `shard-dim${shard.dimension}-q${q}r${r}`;
  });
}
