// @a0n/gnosis/src/void-game-coordinate.ts
/**
 * The Void — Universal Game Coordinate System
 *
 * Every game state is deterministically addressable via a GameCoordinate.
 * Given seed + dimension + hex position + tick, the exact state can be
 * recreated from scratch. The coordinate formalizes the save.
 *
 * Inspired by storms-watch / stars.fish universal atlas tuple.
 * Time is gnosis time (picolorenzos), not human time.
 *
 * @module @a0n/gnosis/void-game-coordinate
 */

import type { GameCoordinate, DimensionIndex } from './void-game-types';
import { DIMENSION_TIME_DILATION } from './void-game-types';

// ── Coordinate Construction ─────────────────────────────────────────

/**
 * Create a valid GameCoordinate.
 * Enforces the cube coordinate constraint: q + r + s = 0.
 */
export function createGameCoordinate(
  seed: number,
  dimension: DimensionIndex,
  hexQ: number,
  hexR: number,
  tick: number
): GameCoordinate {
  const hexS = (-hexQ - hexR) | 0; // | 0 normalizes -0 to 0
  return Object.freeze({
    seed,
    dimension,
    hexQ,
    hexR,
    hexS,
    tick,
  });
}

/**
 * Create the origin coordinate — where every observer begins.
 * Dimension 0 (Bythos), position (0,0,0), tick 0.
 */
export function originCoordinate(seed: number): GameCoordinate {
  return createGameCoordinate(seed, 0, 0, 0, 0);
}

// ── Coordinate Validation ───────────────────────────────────────────

/** Validate a GameCoordinate's cube constraint. */
export function isValidCoordinate(coord: GameCoordinate): boolean {
  return (
    coord.hexQ + coord.hexR + coord.hexS === 0 &&
    coord.dimension >= 0 &&
    coord.dimension <= 12 &&
    coord.seed >= 0 &&
    coord.tick >= 0
  );
}

// ── Coordinate Equality ─────────────────────────────────────────────

/** Check if two coordinates represent the same position. */
export function coordinateEquals(a: GameCoordinate, b: GameCoordinate): boolean {
  return (
    a.seed === b.seed &&
    a.dimension === b.dimension &&
    a.hexQ === b.hexQ &&
    a.hexR === b.hexR &&
    a.hexS === b.hexS &&
    a.tick === b.tick
  );
}

/** Check spatial equality (same position, ignoring tick). */
export function coordinateSpatialEquals(
  a: GameCoordinate,
  b: GameCoordinate
): boolean {
  return (
    a.seed === b.seed &&
    a.dimension === b.dimension &&
    a.hexQ === b.hexQ &&
    a.hexR === b.hexR &&
    a.hexS === b.hexS
  );
}

// ── Coordinate Arithmetic ───────────────────────────────────────────

/** Hex cube distance between two coordinates (must be same seed/dimension). */
export function hexDistance(a: GameCoordinate, b: GameCoordinate): number {
  return Math.max(
    Math.abs(a.hexQ - b.hexQ),
    Math.abs(a.hexR - b.hexR),
    Math.abs(a.hexS - b.hexS)
  );
}

/** The six hex cube direction vectors. */
export const HEX_DIRECTIONS: ReadonlyArray<
  readonly [number, number, number]
> = [
  [1, 0, -1],   // East
  [1, -1, 0],   // NE
  [0, -1, 1],   // NW
  [-1, 0, 1],   // West
  [-1, 1, 0],   // SW
  [0, 1, -1],   // SE
] as const;

/**
 * Move a coordinate in a hex direction.
 * Advances gnosis time by the dimension's time dilation rate.
 */
export function moveCoordinate(
  coord: GameCoordinate,
  directionIndex: number
): GameCoordinate {
  const dir = HEX_DIRECTIONS[directionIndex % 6];
  if (dir === undefined) {
    throw new Error(`Invalid direction index: ${directionIndex}`);
  }
  const [dq, dr] = dir;
  const dilationRate = DIMENSION_TIME_DILATION[coord.dimension];
  return createGameCoordinate(
    coord.seed,
    coord.dimension,
    coord.hexQ + dq,
    coord.hexR + dr,
    coord.tick + dilationRate // Time advances by β₁ picolorenzos per action
  );
}

/** Get all 6 neighbor coordinates from a position. */
export function getNeighbors(coord: GameCoordinate): GameCoordinate[] {
  return HEX_DIRECTIONS.map((_, i) => moveCoordinate(coord, i));
}

/**
 * Change dimension (ascend or descend).
 * Does NOT move spatially — the observer stays at the same hex position.
 */
export function changeDimension(
  coord: GameCoordinate,
  newDimension: DimensionIndex
): GameCoordinate {
  return createGameCoordinate(
    coord.seed,
    newDimension,
    coord.hexQ,
    coord.hexR,
    coord.tick
  );
}

// ── Coordinate Encoding (Bitwise) ───────────────────────────────────

/**
 * Encode a GameCoordinate to a compact binary buffer (32 bytes).
 *
 * Layout:
 *   [0-7]   seed     (float64)
 *   [8]     dimension (uint8)
 *   [9-12]  hexQ     (int32)
 *   [13-16] hexR     (int32)
 *   [17-20] hexS     (int32)
 *   [21-28] tick     (float64)
 *   [29-31] reserved
 */
export function encodeGameCoordinate(coord: GameCoordinate): ArrayBuffer {
  const buffer = new ArrayBuffer(32);
  const view = new DataView(buffer);

  view.setFloat64(0, coord.seed, true);
  view.setUint8(8, coord.dimension);
  view.setInt32(9, coord.hexQ, true);
  view.setInt32(13, coord.hexR, true);
  view.setInt32(17, coord.hexS, true);
  view.setFloat64(21, coord.tick, true);

  return buffer;
}

/**
 * Decode a GameCoordinate from a 32-byte binary buffer.
 */
export function decodeGameCoordinate(buffer: ArrayBuffer): GameCoordinate {
  if (buffer.byteLength < 32) {
    throw new Error(
      `Buffer too small: expected 32 bytes, got ${buffer.byteLength}`
    );
  }

  const view = new DataView(buffer);
  const seed = view.getFloat64(0, true);
  const dimension = view.getUint8(8) as DimensionIndex;
  const hexQ = view.getInt32(9, true);
  const hexR = view.getInt32(13, true);
  const hexS = view.getInt32(17, true);
  const tick = view.getFloat64(21, true);

  // Validate cube constraint
  if (hexQ + hexR + hexS !== 0) {
    throw new Error(
      `Invalid cube coordinates: ${hexQ} + ${hexR} + ${hexS} !== 0`
    );
  }

  return Object.freeze({ seed, dimension, hexQ, hexR, hexS, tick });
}

// ── Coordinate Formatting ───────────────────────────────────────────

/**
 * Human-readable string for a GameCoordinate.
 * Format: "seed:dim(q,r,s)@tick"
 */
export function coordinateToString(coord: GameCoordinate): string {
  return `${coord.seed}:d${coord.dimension}(${coord.hexQ},${coord.hexR},${coord.hexS})@${coord.tick.toFixed(3)}pLo`;
}

// ── Deterministic Seed Derivation ───────────────────────────────────

/**
 * Derive a deterministic sub-seed from a coordinate.
 * Used for procedural generation: given seed + position,
 * the tile content is always the same.
 *
 * Uses a simple but effective hash mixing function.
 */
export function coordinateToTileSeed(coord: GameCoordinate): number {
  // Mix seed with position using FNV-1a-like hash
  let hash = 2166136261; // FNV offset basis
  hash = Math.imul(hash ^ (coord.seed | 0), 16777619);
  hash = Math.imul(hash ^ coord.dimension, 16777619);
  hash = Math.imul(hash ^ coord.hexQ, 16777619);
  hash = Math.imul(hash ^ coord.hexR, 16777619);
  hash = Math.imul(hash ^ coord.hexS, 16777619);
  // Normalize to [0, 1)
  return (hash >>> 0) / 4294967296;
}
