// @a0n/gnosis/src/void-channel.ts
/**
 * void-channel.ts -- Shannon channel capacity of void walking
 *
 * The void boundary is a noisy channel: the walker sends choices,
 * the environment responds with rejections, and the complement
 * distribution is the posterior. The channel capacity bounds
 * how fast the walker can learn.
 *
 * FOLD is an information bottleneck: it compresses multiple branches
 * into one, losing at least H(branches) - H(survivor) bits.
 *
 * VENT is a rate-distortion operator: it discards information to
 * stay within a bit budget, and the distortion is the KL divergence
 * from the true posterior.
 */

import {
  type VoidBoundary,
  complementDistribution,
  shannonEntropy,
} from './void';
import {
  klDivergence,
  jensenShannonDivergence,
} from './information-geometry';

// ============================================================================
// Channel capacity
// ============================================================================

/**
 * Mutual information I(X;Y) between choice X and observation Y.
 * For void walking: X = chosen dimension, Y = rejection signal.
 * I(X;Y) = H(X) - H(X|Y) = H(complement) - H(complement | rejection).
 *
 * Upper bounded by log(n) (n = number of dimensions).
 * Lower bounded by 0 (no information gained).
 */
export function mutualInformation(
  priorDist: number[],
  posteriorDist: number[]
): number {
  // I(X;Y) = D_KL(posterior || prior) when Y is observed
  return klDivergence(posteriorDist, priorDist);
}

/**
 * Channel capacity: maximum mutual information over all input distributions.
 * For the void channel with n dimensions, capacity = log(n).
 * This is achieved when the complement distribution is uniform.
 */
export function channelCapacity(numDimensions: number): number {
  return Math.log(numDimensions);
}

/**
 * Information gain per void walk step.
 * Before step: prior = complementDistribution(boundary, eta)
 * After step: posterior = complementDistribution(updated_boundary, eta)
 * Gain = D_KL(posterior || prior)
 */
export function informationGainPerStep(
  boundaryBefore: VoidBoundary,
  boundaryAfter: VoidBoundary,
  eta: number
): number {
  const prior = complementDistribution(boundaryBefore, eta);
  const posterior = complementDistribution(boundaryAfter, eta);
  return klDivergence(posterior, prior);
}

/**
 * Learning rate: cumulative information gain / number of steps.
 * Bounded by channel capacity / step.
 */
export function learningRate(totalInfoGain: number, steps: number): number {
  return steps > 0 ? totalInfoGain / steps : 0;
}

// ============================================================================
// Information bottleneck at FOLD
// ============================================================================

/**
 * FOLD compresses k branches into 1.
 * Minimum information loss = H(branches) - H(survivor).
 * When branches have equal probability: loss = log(k).
 */
export function foldInformationLoss(
  branchDistribution: number[],
  survivorProbability: number
): number {
  const branchEntropy = shannonEntropy(branchDistribution);
  // Survivor carries at most -log(p_survivor) bits
  const survivorInfo =
    survivorProbability > 0 ? -Math.log(survivorProbability) : 0;
  return Math.max(0, branchEntropy - survivorInfo);
}

/**
 * Tight FOLD loss for k equal branches: exactly log(k) bits.
 */
export function foldLossEqualBranches(k: number): number {
  return k > 1 ? Math.log(k) : 0;
}

/**
 * Information bottleneck bound: FOLD cannot preserve more than
 * I(survivor; branches) bits of information about the branches.
 */
export function informationBottleneckBound(
  branchDistribution: number[]
): number {
  return shannonEntropy(branchDistribution);
}

// ============================================================================
// Rate-distortion at VENT
// ============================================================================

/**
 * VENT discards a dimension, incurring distortion.
 * Distortion = KL divergence from the true (pre-vent) posterior
 * to the reduced (post-vent) posterior.
 *
 * Rate = bits saved by venting.
 * Distortion = information lost about the vented dimension.
 */
export function ventRateDistortion(
  preDist: number[],
  postDist: number[]
): { rate: number; distortion: number } {
  const rate = shannonEntropy(preDist) - shannonEntropy(postDist);
  const distortion = klDivergence(preDist, postDist);
  return { rate: Math.max(0, rate), distortion };
}

/**
 * Minimum distortion for a given rate budget.
 * By the rate-distortion theorem: D(R) = H - R for R < H.
 */
export function minimumDistortion(entropy: number, rateBudget: number): number {
  return Math.max(0, entropy - rateBudget);
}

// ============================================================================
// Mixing time bounds from topology structure
// ============================================================================

/**
 * Mixing time: number of steps until the walker's distribution
 * is within epsilon of the stationary distribution.
 *
 * T_mix(ε) ≤ (1/gap) * log(1/ε)
 * where gap = spectral gap of the transition kernel.
 */
export function mixingTimeBound(
  spectralGap: number,
  epsilon: number = 0.01
): number {
  if (spectralGap <= 0) return Infinity;
  return Math.ceil(Math.log(1 / epsilon) / spectralGap);
}

/**
 * Spectral gap lower bound from topology structure.
 * For a connected graph with n nodes and diameter D:
 * gap ≥ 1/(n*D)
 */
export function spectralGapLowerBound(
  numNodes: number,
  diameter: number
): number {
  return numNodes > 0 && diameter > 0 ? 1 / (numNodes * diameter) : 0;
}

/**
 * Mixing time upper bound from graph structure.
 * T_mix ≤ n * D * log(1/ε)
 */
export function mixingTimeFromGraph(
  numNodes: number,
  diameter: number,
  epsilon: number = 0.01
): number {
  return Math.ceil(numNodes * diameter * Math.log(1 / epsilon));
}

// ============================================================================
// Concentration inequalities for complement distribution
// ============================================================================

/**
 * Hoeffding bound: after t steps, the empirical complement distribution
 * concentrates around the true stationary distribution.
 *
 * P(|p_hat - p| > δ) ≤ 2 exp(-2tδ²)
 */
export function hoeffdingBound(steps: number, delta: number): number {
  return 2 * Math.exp(-2 * steps * delta * delta);
}

/**
 * Minimum steps for concentration within delta with probability 1-alpha.
 * t ≥ log(2/alpha) / (2δ²)
 */
export function minimumStepsForConcentration(
  delta: number,
  alpha: number = 0.05
): number {
  return Math.ceil(Math.log(2 / alpha) / (2 * delta * delta));
}

/**
 * The complement distribution concentrates exponentially fast.
 * After t steps, the total variation distance to stationary is at most
 * (1 - gap)^t where gap is the spectral gap.
 */
export function concentrationRate(spectralGap: number, steps: number): number {
  return Math.pow(1 - spectralGap, steps);
}
