// @a0n/gnosis/src/void-topology.ts
/**
 * void-topology.ts -- The bridge where topology meets void
 *
 * Any GraphAST becomes a walkable void boundary. The topology's nodes
 * are dimensions. Edge types define flow semantics. The engine walks
 * the topology's void boundary using the c0-c3 loop, and the complement
 * distribution guides execution decisions.
 *
 * This is where gnosis eats itself: the topology that describes void
 * walking is itself walked by the void walker during execution.
 */

import type { GraphAST, ASTEdge } from './betty/compiler';
import {
  type VoidBoundary,
  type Walker,
  type BoundaryStack,
  type TimescaleBoundary,
  type Measurement,
  type Resonance,
  type Gait,
  createVoidBoundary,
  updateVoidBoundary,
  complementDistribution,
//   sampleComplement,
//   measure,
  createWalker,
  createTimescaleBoundary,
  createBoundaryStack,
  createResonance,
//   flattenStack,
//   measureStack,
  c0Choose,
  c0Update,
  c1Measure,
  c2c3Adapt,
//   selectGait,
  GAIT_DEPTH,
} from './void';

// ============================================================================
// VoidTopology -- any GraphAST as a walkable void boundary
// ============================================================================

/**
 * A VoidTopology wraps a GraphAST and maintains a void boundary
 * where each node is a dimension. As the engine executes nodes,
 * the boundary accumulates. The complement distribution tells
 * the engine which paths are freshest -- least executed, most
 * promising.
 */
export interface VoidTopology {
  /** The underlying graph */
  ast: GraphAST;
  /** Node-indexed void boundary (one dimension per node) */
  boundary: VoidBoundary;
  /** Map from node ID to boundary dimension index */
  nodeIndex: Map<string, number>;
  /** Reverse map: dimension index to node ID */
  indexToNode: string[];
  /** The walker that navigates this topology */
  walker: Walker;
  /** Execution history: which nodes were chosen and what they returned */
  history: ExecutionStep[];
}

export interface ExecutionStep {
  nodeId: string;
  dimensionIdx: number;
  gait: Gait;
  measurement: Measurement;
  /** Whether the node's handler succeeded */
  succeeded: boolean;
  /** Rejection magnitude (higher = more void accumulated) */
  rejection: number;
}

/**
 * Create a VoidTopology from any GraphAST.
 * Each node becomes a dimension in the void boundary.
 */
export function createVoidTopology(ast: GraphAST): VoidTopology {
  const nodeIds = Array.from(ast.nodes.keys());
  const nodeIndex = new Map<string, number>();
  nodeIds.forEach((id, i) => nodeIndex.set(id, i));

  const boundary = createVoidBoundary(nodeIds.length);
  const walker = createWalker(boundary);

  return {
    ast,
    boundary,
    nodeIndex,
    indexToNode: nodeIds,
    walker,
    history: [],
  };
}

// ============================================================================
// Void-guided execution decisions
// ============================================================================

/**
 * At a FORK point, rank targets by complement distribution.
 * Nodes with least accumulated void get highest probability.
 * Returns targets sorted by complement weight (highest first).
 */
export function rankForkTargets(
  vt: VoidTopology,
  targetIds: string[]
): string[] {
  const dist = complementDistribution(vt.boundary, vt.walker.eta);
  const ranked = targetIds
    .map((id) => ({
      id,
      weight: dist[vt.nodeIndex.get(id) ?? 0] ?? 0,
    }))
    // Law 2 (strict ordering): primary sort by weight (antitone), secondary by id for determinism
    .sort((a, b) => b.weight - a.weight || a.id.localeCompare(b.id));
  return ranked.map((r) => r.id);
}

/**
 * Result of branch selection at a FORK.
 * The paradigm (THM-SERVER-OPTIMALITY) requires every FORK to account
 * for all branches: active branches run, vented branches are explicitly
 * discarded. No branch may be silently ignored.
 */
export interface BranchSelection {
  /** Branches selected for execution (by gait depth) */
  active: string[];
  /** Branches explicitly vented (not selected, tracked in void boundary) */
  vented: string[];
}

/**
 * At a FORK, use gait to decide how many branches to run.
 * Stand: 1 (sequential). Trot: 1 (best complement).
 * Canter: up to 4. Gallop: all of them.
 *
 * Returns both active and vented branches. The paradigm requires
 * explicit VENT accounting -- vented branches contribute to the void
 * boundary (THM-VOID-BOUNDARY-MEASURABLE) and drive the complement
 * distribution for future decisions (THM-VOID-GRADIENT).
 */
export function selectActiveBranches(
  vt: VoidTopology,
  targetIds: string[]
): BranchSelection {
  const ranked = rankForkTargets(vt, targetIds);
  const depth = GAIT_DEPTH[vt.walker.gait];
  const cutoff = Math.min(depth, ranked.length);
  const active = ranked.slice(0, cutoff);
  const vented = ranked.slice(cutoff);

  // Vented branches accumulate in the void boundary -- they are not
  // lost but become the negative information that guides future forks.
  for (const ventedId of vented) {
    const dimIndex = vt.nodeIndex.get(ventedId);
    if (dimIndex !== undefined) {
      updateVoidBoundary(vt.boundary, dimIndex);
    }
  }

  return { active, vented };
}

/**
 * At a SUPERPOSE, replace Math.random() with complement-weighted sampling.
 * The topology's execution history shapes which amplitudes survive.
 */
export function sampleSuperposition(
  vt: VoidTopology,
  targetIds: string[],
  rng: () => number
): string[] {
  const dist = complementDistribution(vt.boundary, vt.walker.eta);
  const selected: string[] = [];
  for (const id of targetIds) {
    const idx = vt.nodeIndex.get(id);
    if (idx !== undefined) {
      const weight = dist[idx];
      if (weight !== undefined && rng() < weight * targetIds.length) {
        selected.push(id);
      }
    }
  }
  // Always keep at least one
  if (selected.length === 0) {
    const firstTarget = targetIds[0];
    if (firstTarget !== undefined) {
      selected.push(firstTarget);
    }
  }
  return selected;
}

/**
 * After a node executes, update the void boundary.
 * Success = small rejection (the node worked, slight accumulation).
 * Failure = large rejection (the node failed, strong avoidance signal).
 * Slow = medium rejection (the node was slow, moderate avoidance).
 */
export function recordExecution(
  vt: VoidTopology,
  nodeId: string,
  succeeded: boolean,
  durationMs?: number
): void {
  const idx = vt.nodeIndex.get(nodeId);
  if (idx === undefined) return;

  // Rejection magnitude: failed nodes accumulate more void
  let rejection: number;
  if (!succeeded) {
    rejection = 3.0; // Strong avoidance
  } else if (durationMs !== undefined && durationMs > 1000) {
    rejection = 1.0 + Math.log10(durationMs / 1000); // Slow penalty
  } else {
    rejection = 0.5; // Minimal accumulation for success
  }

  c0Update(vt.walker, idx, rejection);
  c2c3Adapt(vt.walker);

  const m = c1Measure(vt.walker);
  vt.history.push({
    nodeId,
    dimensionIdx: idx,
    gait: vt.walker.gait,
    measurement: m,
    succeeded,
    rejection,
  });
}

/**
 * Choose the next node to visit when the topology has multiple
 * possible starting points or when exploring freely.
 */
export function chooseNext(vt: VoidTopology, rng: () => number): string {
  const idx = c0Choose(vt.walker, rng);
  const nextNode = vt.indexToNode[idx];
  if (nextNode === undefined) {
    throw new Error(`[chooseNext] missing node for index ${idx}`);
  }
  return nextNode;
}

// ============================================================================
// Topology as BoundaryStack -- layered execution
// ============================================================================

/**
 * Decompose a topology into a BoundaryStack by grouping nodes
 * into layers based on their topological depth (distance from roots).
 *
 * This is how a flat topology becomes a layered void boundary system:
 * root nodes = deepest timescale (they constrain everything downstream),
 * leaf nodes = shallowest timescale (most reactive to execution history).
 */
export function topologyToBoundaryStack(vt: VoidTopology): BoundaryStack {
  const ast = vt.ast;

  // Compute topological depth for each node
  const allTargetIds = new Set<string>();
  ast.edges.forEach((e) =>
    e.targetIds.forEach((id) => allTargetIds.add(id.trim()))
  );
  const roots = Array.from(ast.nodes.keys()).filter(
    (id) => !allTargetIds.has(id.trim())
  );

  const depth = new Map<string, number>();
  const queue = roots.map((id) => ({ id, d: 0 }));
  while (queue.length > 0) {
    const { id, d } = queue.shift()!;
    if (depth.has(id) && depth.get(id)! <= d) continue;
    depth.set(id, d);
    const outEdges = ast.edges.filter((e) =>
      e.sourceIds.map((s) => s.trim()).includes(id)
    );
    for (const edge of outEdges) {
      for (const targetId of edge.targetIds) {
        queue.push({ id: targetId.trim(), d: d + 1 });
      }
    }
  }

  // Group nodes by depth
  const maxDepth = Math.max(...Array.from(depth.values()), 0);
  const layerNodes: string[][] = Array.from({ length: maxDepth + 1 }, () => []);
  for (const [nodeId, d] of depth) {
    const layer = layerNodes[d];
    if (layer === undefined) {
      throw new Error(`Missing layer bucket for depth ${d}`);
    }
    layer.push(nodeId);
  }
  // Include orphan nodes (not reachable from roots)
  for (const nodeId of ast.nodes.keys()) {
    if (!depth.has(nodeId)) {
      const fallbackLayer = layerNodes[maxDepth];
      if (fallbackLayer === undefined) {
        throw new Error('Missing fallback layer');
      }
      fallbackLayer.push(nodeId);
    }
  }

  // Timescale mapping: deeper layers = slower timescale
  const timescales: Array<
    'instant' | 'seconds' | 'minutes' | 'hours' | 'days'
  > = ['days', 'hours', 'minutes', 'seconds', 'instant'];

  const layers: TimescaleBoundary[] = layerNodes
    .filter((nodes) => nodes.length > 0)
    .map((nodes, i) => {
      const ts = timescales[Math.min(i, timescales.length - 1)];
      if (ts === undefined) {
        throw new Error(`Missing timescale for depth layer ${i}`);
      }
      return createTimescaleBoundary(`depth-${i}`, ts, nodes.length, nodes);
    });

  // Resonance: edges that skip depths create cross-layer coupling
  const resonances: Resonance[] = [];
  for (const edge of ast.edges) {
    for (const sourceId of edge.sourceIds) {
      for (const targetId of edge.targetIds) {
        const sourceDepth = depth.get(sourceId.trim());
        const targetDepth = depth.get(targetId.trim());
        if (
          sourceDepth !== undefined &&
          targetDepth !== undefined &&
          Math.abs(targetDepth - sourceDepth) > 1
        ) {
          resonances.push(
            createResonance(
              Math.min(sourceDepth, targetDepth),
              Math.max(sourceDepth, targetDepth),
              0.05,
              `${edge.type}: ${sourceId} -> ${targetId}`
            )
          );
        }
      }
    }
  }

  return createBoundaryStack(
    `topology:${ast.nodes.size}n/${ast.edges.length}e`,
    layers,
    resonances
  );
}

// ============================================================================
// Self-measurement -- the topology measures itself
// ============================================================================

/**
 * The topology measures its own execution state.
 * This is the self-referential moment: the measurement operators
 * (entropy, kurtosis, gini, inverseBule) that the topology defines
 * are applied to the topology's own void boundary.
 */
export function measureTopology(vt: VoidTopology): {
  /** Per-node measurement of the void boundary */
  boundary: Measurement;
  /** Current walker gait */
  gait: Gait;
  /** How many unique nodes have been visited */
  coverage: number;
  /** Total execution steps */
  steps: number;
  /** Nodes ranked by complement weight (freshest first) */
  freshest: string[];
  /** Nodes ranked by void accumulation (most avoided first) */
  mostAvoided: string[];
} {
  const m = c1Measure(vt.walker);
  const dist = complementDistribution(vt.boundary, vt.walker.eta);

  // Rank by complement (freshest = highest complement weight)
  const ranked = vt.indexToNode
    .map((id, i) => {
      const complement = dist[i];
      const voidCount = vt.boundary.counts[i];
      if (id === undefined || complement === undefined || voidCount === undefined) {
        return null;
      }
      return { id, complement, void: voidCount };
    })
    .filter(
      (
        value
      ): value is {
        id: string;
        complement: number;
        void: number;
      } => value !== null
    )
    .sort((a, b) => b.complement - a.complement);

  const visitedNodes = new Set(vt.history.map((h) => h.nodeId));

  return {
    boundary: m,
    gait: vt.walker.gait,
    coverage: visitedNodes.size / vt.indexToNode.length,
    steps: vt.walker.steps,
    freshest: ranked.map((r) => r.id),
    mostAvoided: [...ranked].sort((a, b) => b.void - a.void).map((r) => r.id),
  };
}

// ============================================================================
// Edge-type semantics as void operations
// ============================================================================

/**
 * Map edge types to void operations.
 * This is the semantic bridge: each .gg edge type has a precise
 * meaning in void walking terms.
 */
export const EDGE_VOID_SEMANTICS: Record<
  string,
  {
    /** How much void accumulates on traversal */
    baseMagnitude: number;
    /** Whether this edge creates new void dimensions */
    spawns: boolean;
    /** Whether this edge collapses void dimensions */
    collapses: boolean;
    /** Description in void walking terms */
    voidMeaning: string;
  }
> = {
  PROCESS: {
    baseMagnitude: 0.5,
    spawns: false,
    collapses: false,
    voidMeaning: 'Sequential step -- minimal void accumulation',
  },
  FORK: {
    baseMagnitude: 0.1,
    spawns: true,
    collapses: false,
    voidMeaning: 'Boundary expands -- new void dimensions open',
  },
  RACE: {
    baseMagnitude: 2.0,
    spawns: false,
    collapses: true,
    voidMeaning: 'Losers accumulate void -- only the complement survives',
  },
  FOLD: {
    baseMagnitude: 1.0,
    spawns: false,
    collapses: true,
    voidMeaning: 'All paths merge -- void boundary contracts',
  },
  COLLAPSE: {
    baseMagnitude: 1.5,
    spawns: false,
    collapses: true,
    voidMeaning: 'Forced measurement -- superposition resolves',
  },
  VENT: {
    baseMagnitude: 5.0,
    spawns: false,
    collapses: false,
    voidMeaning: 'Rejection ejected -- maximum void at this dimension',
  },
  EVOLVE: {
    baseMagnitude: 0.3,
    spawns: true,
    collapses: false,
    voidMeaning: 'Reynolds-modulated expansion -- boundary breathes',
  },
  SUPERPOSE: {
    baseMagnitude: 0.0,
    spawns: true,
    collapses: false,
    voidMeaning: 'Probabilistic boundary -- complement formalizes the amplitude',
  },
  ENTANGLE: {
    baseMagnitude: 0.0,
    spawns: true,
    collapses: false,
    voidMeaning: 'Shared void -- boundaries couple across branches',
  },
  OBSERVE: {
    baseMagnitude: 1.0,
    spawns: false,
    collapses: true,
    voidMeaning: 'Measurement collapses the complement distribution',
  },
  TUNNEL: {
    baseMagnitude: 0.0,
    spawns: false,
    collapses: true,
    voidMeaning: 'Escape through the boundary -- void bypassed',
  },
  LAMINAR: {
    baseMagnitude: 0.5,
    spawns: false,
    collapses: false,
    voidMeaning: 'Self-contained void loop -- beta-neutral',
  },
  HALT: {
    baseMagnitude: 0.0,
    spawns: false,
    collapses: false,
    voidMeaning: 'Walker pauses -- boundary freezes',
  },
  MEASURE: {
    baseMagnitude: 0.0,
    spawns: false,
    collapses: false,
    voidMeaning: 'Self-measurement -- the topology observes itself',
  },
};

/**
 * After an edge is traversed, apply its void semantics.
 */
export function recordEdgeTraversal(
  vt: VoidTopology,
  edge: ASTEdge,
  succeeded: boolean
): void {
  const semantics = EDGE_VOID_SEMANTICS[edge.type];
  if (semantics === undefined) {
    return;
  }
  const multiplier = succeeded ? 1.0 : 2.0;

  for (const targetId of edge.targetIds) {
    const idx = vt.nodeIndex.get(targetId.trim());
    if (idx !== undefined) {
      updateVoidBoundary(
        vt.boundary,
        idx,
        semantics.baseMagnitude * multiplier
      );
    }
  }
}
