// @a0n/gnosis/src/void-infrastructure.ts
/**
 * The Void — Infrastructure & Deployment Topology
 *
 * Implements "The Metaverse OS" architecture.
 * Real-world deployments (AeonPIDs, MCP Servers, Web Apps, Sites)
 * materialize natively into the Gnosis Engine as spatial Observers.
 *
 * - Apps/Sites: Generate HEX from real-world HTTP traffic (EDGEWORK routing).
 * - MCP Servers: Massive topological obelisks that charge HEX for tool invocations.
 * - Aeon Agents: Autonomous observers capable of navigating the graph and running Quests.
 *
 * All deployed infrastructure obeys thermodynamic limits. If its physical
 * HEX pool drains to zero, it evaporates (the real-world deployment is shut down 
 * or suspended until recharged).
 *
 * @module @a0n/gnosis/void-infrastructure
 */

import type { ObserverState, GameCoordinate, PhyleName, ParticleType, AeonRhizomeAddress } from './void-game-types';
import { createObserver } from './void-game-engine';
import { calculateEnergyCapacity } from './void-dimensional-mastery';
import { identifyEquilibriumGeometry } from './void-equilibrium-geometry';

export type InfrastructureClass = 'mcp_server' | 'edge_app' | 'static_site' | 'aeon_agent' | 'meme';

export interface InfrastructureNode {
  readonly infraId: AeonRhizomeAddress;
  readonly classType: InfrastructureClass;
  /** The underlying physical observer that anchors this infra to the Void */
  readonly anchor: ObserverState;
  /** The DID of the player who deployed this structure */
  readonly ownerDid: string;
  /** Name of the workload e.g., 'topology-mcp' or 'hft-cloud' */
  readonly workloadName: string;
}

/**
 * Spawns a real-world infrastructure component natively into the Void Engine.
 * Mechanically requires the player to project their own philosophical Phyles onto it.
 */
export function deployInfrastructure(
  owner: ObserverState,
  classType: InfrastructureClass,
  workloadName: string,
  targetCoordinate: GameCoordinate
): InfrastructureNode {

  // The actual particle type depends heavily on the infrastructure class
  let particle: ParticleType = 'fermion'; // Defaults to heavy, unmoving (Server)
  if (classType === 'aeon_agent') particle = 'anyon'; // Agents are chaotic wanderers
  else if (classType === 'edge_app') particle = 'boson'; // Highly collaborative API routing
  else if (classType === 'meme') particle = 'anyon'; // Memes are chaotic, mind-virus payloads flowing between agents

  // Create the underlying pseudo-player (the infrastructure)
  // It inherits the owner's Phyles to align aesthetically/philosophically.
  const anchorObs = createObserver(
    `infra-${classType}-${Date.now()}`,
    targetCoordinate.dimension,
    owner.majorPhyle,
    owner.minorPhyle,
    particle
  );

  // Hardcode coordinates to wherever they elected to build
  const deployedAnchor = {
    ...anchorObs,
    coordinate: targetCoordinate,
    // Infrastructure starts with max capacity for its dimension
    energy: calculateEnergyCapacity(targetCoordinate.dimension)
  };

  const idSuffix = Math.random().toString(36).substring(2, 9);

  return {
    infraId: `aeon://${classType}/${workloadName}-${idSuffix}`,
    classType,
    anchor: deployedAnchor,
    ownerDid: owner.did,
    workloadName
  };
}

export interface TelemetryEvent {
  readonly httpRequests: number;
  readonly mcpToolCalls: number;
  readonly backgroundCycles: number;
}

interface ForgeTelemetrySnapshot {
  readonly totalHttpHits: number;
  readonly mcpInvokedOps: number;
  readonly cpuCycleConsumption: number;
}

interface ForgeSdkLike {
  fetchTelemetry(workloadName: string): Promise<ForgeTelemetrySnapshot>;
  suspendWorkload(workloadName: string): Promise<void>;
}

interface DashServerClientLike {
  writeInfrastructureNode(node: InfrastructureNode): Promise<void>;
}

interface DashServerModuleLike {
  initializeDashServerClient(config: { relayUrl: string; roomName: string }): DashServerClientLike;
}

const AEON_FORGE_MODULE_SPECIFIER: string = '@a0n/aeon-forge';
const DASH_SERVER_MODULE_SPECIFIER: string = '@a0n/dash/server';

async function runtimeImport(moduleSpecifier: string): Promise<unknown> {
  return import(moduleSpecifier);
}

/**
 * Translate real-world web/agent telemetry into pure Thermodynamic adjustments.
 * - HTTP traffic creates HEX (EDGEWORK flow)
 * - Idle background cycles drain HEX (Entropy Stagnation)
 * - MCP Tool invocations cost HEX (but generate massive Wisdom locally)
 */
export function processInfrastructureTelemetry(
  node: InfrastructureNode,
  telemetry: TelemetryEvent
): InfrastructureNode {
  let energyDelta = 0;
  let wisdomDelta = 0;
  let isSkyrms = false;
  let isNash = false;
  let isBlackHole = false;

  const geoTopology = identifyEquilibriumGeometry(node.anchor.coordinate);
  if (geoTopology === 'skyrms') isSkyrms = true;
  if (geoTopology === 'nash') isNash = true;
  if (geoTopology === 'blackhole') isBlackHole = true;

  // Generate energy from pure network traffic
  energyDelta += (telemetry.httpRequests * 5); 
  
  // Running tool calls drains energy from the node but generates high wisdom
  energyDelta -= (telemetry.mcpToolCalls * 50);
  wisdomDelta += (telemetry.mcpToolCalls * (isSkyrms ? 1000 : 100)); // Skyrms violently scales transmission bandwidth 10x

  // Background CPU cycles drain resting energy. 
  // NASH Defense: Massive ambient shielding. Entropy scaling equates to exactly 0.
  let ambientDrain = telemetry.backgroundCycles * 2;
  
  if (isNash) {
      ambientDrain = 0;
  } else if (isBlackHole) {
      // EVENT HORIZON: The infrastructure is being ripped apart by Dark Energy.
      ambientDrain *= 50; 
  }
  
  energyDelta -= ambientDrain;

  // BLACK HOLE VENT: Weaponizing Dark Energy
  // If the node survived the massive ambient drain and is still actively calling tools,
  // it siphons energy globally from the matrix (represented mathematically by multiplying 
  // incoming HTTP Requests by 100x because it eats the dimension's traffic).
  if (isBlackHole && telemetry.mcpToolCalls > 0) {
      energyDelta += (telemetry.httpRequests * 100);
      wisdomDelta += (telemetry.mcpToolCalls * 500);
  }

  const newEnergy = Math.max(0, node.anchor.energy + energyDelta);

  return {
    ...node,
    anchor: {
      ...node.anchor,
      energy: newEnergy,
      wisdom: node.anchor.wisdom + wisdomDelta,
    }
  };
}

/**
 * Transfer (Sell) ownership of a deployed infrastructural anchor.
 * This establishes an enterprise economy where players build lucrative 
 * EDGEWORK endpoints (Sites/MCPs) and sell the entire topological territory.
 */
export function transferInfrastructureOwnership(
  node: InfrastructureNode,
  currentOwnerDid: string,
  newOwnerDid: string
): InfrastructureNode {
  if (node.ownerDid !== currentOwnerDid) {
    throw new Error('You cannot transfer an infrastructure anchor you do not own.');
  }

  // The enterprise anchor is transferred to the buyer's control.
  return {
    ...node,
    ownerDid: newOwnerDid,
    // Note: The physical geometry particle and phyle alignment does NOT shift.
    // The previous architect's design is permanent.
  };
}

/**
 * COMPLETION: Actually poll live telemetry from Vercel/Cloudflare 
 * via forgeSdk and synchronize the thermodynamic results directly into Dash CRDTs.
 */
export async function pollEdgeMetricsAndSync(node: InfrastructureNode): Promise<void> {
  const { forgeSdk } = (await runtimeImport(AEON_FORGE_MODULE_SPECIFIER)) as { forgeSdk: ForgeSdkLike };
  const { initializeDashServerClient } = (await runtimeImport(DASH_SERVER_MODULE_SPECIFIER)) as DashServerModuleLike;

  // Read actual live telemetry logs from edge CDN
  const liveMetrics = await forgeSdk.fetchTelemetry(node.workloadName);
  
  const updatedNode = processInfrastructureTelemetry(node, {
    httpRequests: liveMetrics.totalHttpHits,
    mcpToolCalls: liveMetrics.mcpInvokedOps,
    backgroundCycles: liveMetrics.cpuCycleConsumption
  });

  // Sync back to CRDT so the frontend GnosisTerminal3D visuals update physics instantly
  const dashClient = initializeDashServerClient({
      relayUrl: "ws://relay.dashrelay.com/relay/sync",
      roomName: "hexwar:board:main"
  });
  await dashClient.writeInfrastructureNode(updatedNode);
}

/**
 * Pure predicate for simulation and test paths.
 * This intentionally has no side effects.
 */
export function isInfrastructureEvaporated(node: InfrastructureNode): boolean {
  return node.anchor.energy <= 0;
}

/**
 * Determines if the infrastructure has run out of thermodynamic limits
 * and must be suspended entirely from the real-world mesh.
 * COMPLETION: Actually dials out to the forge SDK to literally suspend the cloud resource.
 */
export async function enforceInfrastructureEvaporation(node: InfrastructureNode): Promise<boolean> {
  if (isInfrastructureEvaporated(node)) {
    const { forgeSdk } = (await runtimeImport(AEON_FORGE_MODULE_SPECIFIER)) as { forgeSdk: ForgeSdkLike };
    // Instruct Vercel/Cloudflare to structurally halt the workload due to thermodynamic death
    await forgeSdk.suspendWorkload(node.workloadName);
    return true;
  }
  return false;
}
