// @a0n/gnosis/src/void-user-scripting.ts
/**
 * The Void — Execution Layer (Compute Topologies)
 *
 * Implements the sandbox where observers programmatically execute topology
 * programs as persistent "Compute Topologies".
 *
 * First-class language support:
 * - `.gg` (legacy/default)
 * - `.gnot` (validated + manifest-carrying)
 *
 * "Compute topologies - no tropes."
 *
 * Thermodynamic Halting:
 * Every syntax node (tick) evaluated physically costs Observer Energy (HEX).
 *
 * Topology Economy:
 * Users can forge, trade, and evaluate these topological artifacts. A well-written
 * mathematical algorithm (e.g. an efficient pathfinding mesh) becomes a highly
 * valuable geometric asset in the void.
 *
 * @module @a0n/gnosis/void-user-scripting
 */

import type { ObserverState, AeonRhizomeAddress } from './void-game-types';
import { DIMENSION_TIME_DILATION } from './void-game-types';

export type TopologyASTNode =
  | { type: 'Loop'; body: TopologyASTNode[]; iterations: number }
  | { type: 'Action'; actionType: 'move' | 'rest' | 'affirmStructure' | 'entwine'; target?: string }
  | { type: 'Math'; operation: 'add' | 'sub' | 'mul' | 'div' }
  | { type: 'VariableAssignment' };

export type TopologySourceLanguage = 'gg' | 'gnot';

export interface GnotTopologyManifest {
  readonly appId: string;
  readonly version: string;
  readonly hasDeployBlock: boolean;
  readonly declaredNamespaces: string[];
}

export interface GnotValidationSuccess {
  readonly ok: true;
  readonly normalizedSource: string;
  readonly manifest: GnotTopologyManifest;
}

export interface GnotValidationFailure {
  readonly ok: false;
  readonly errors: string[];
}

export type GnotValidationResult =
  | GnotValidationSuccess
  | GnotValidationFailure;

export type GnotValidator = (source: string) => GnotValidationResult;

export interface ForgeTopologyOptions {
  readonly sourceLanguage?: TopologySourceLanguage;
  readonly gnotValidator?: GnotValidator;
}

export interface ComputeTopology {
  readonly id: AeonRhizomeAddress;
  readonly programName: string;
  readonly authorDid: string; // The original creator
  readonly ownerDid: string; // The current bearer
  readonly sourceLanguage: TopologySourceLanguage;
  readonly sourceCode: string; // Canonical source for the declared language
  readonly sourceGGCode: string; // Backwards-compat alias retained for existing callers
  readonly sourceDigest: string;
  readonly gnotManifest?: GnotTopologyManifest;
  readonly ast: TopologyASTNode[]; // Compiled execution tree
  readonly currentTickIndex: number;
}

const GNOT_HEADER_PATTERN = /^gnot\s+([A-Za-z0-9._-]+)\s+v([0-9]+(?:\.[0-9]+)*)$/;

function normalizeSource(source: string): string {
  const normalized = source
    .replace(/\r\n/g, '\n')
    .split('\n')
    .map((line) => line.replace(/[ \t]+$/u, ''))
    .join('\n')
    .trim();
  return normalized.length > 0 ? `${normalized}\n` : '';
}

function digestSource(source: string): string {
  // Deterministic, runtime-portable digest for identity/fingerprinting.
  let hash = 0x811c9dc5;
  for (let i = 0; i < source.length; i += 1) {
    hash ^= source.charCodeAt(i);
    hash = Math.imul(hash, 0x01000193);
  }
  const unsigned = hash >>> 0;
  return `fnv1a32:${unsigned.toString(16).padStart(8, '0')}:${source.length}`;
}

function significantLines(source: string): string[] {
  return source
    .split('\n')
    .map((line) => line.trim())
    .filter((line) => line.length > 0 && !line.startsWith('#'));
}

function parseNamespaces(lines: string[]): string[] {
  const declarations = lines
    .filter((line) => line.startsWith('using '))
    .flatMap((line) =>
      line
        .slice('using '.length)
        .split(/[,\s]+/u)
        .map((value) => value.trim())
        .filter((value) => value.length > 0)
    );
  return [...new Set(declarations)];
}

/**
 * Handles the gnosis validate Gnot Topology Source workflow.
 */
export function validateGnotTopologySource(source: string): GnotValidationResult {
  const normalized = normalizeSource(source);
  if (normalized.length === 0) {
    return {
      ok: false,
      errors: ['gnot source is empty'],
    };
  }

  const lines = significantLines(normalized);
  if (lines.length === 0) {
    return {
      ok: false,
      errors: ['gnot source does not contain executable lines'],
    };
  }

  const header = lines[0] ?? '';
  const headerMatch = header.match(GNOT_HEADER_PATTERN);
  if (!headerMatch) {
    return {
      ok: false,
      errors: [
        'gnot source must begin with a header like "gnot <app-id> v<version>"',
      ],
    };
  }

  const hasProgramBody = lines.slice(1).some((line) => {
    return (
      line.includes(' is ') ||
      line.startsWith('deploy ') ||
      line.startsWith('stack ')
    );
  });

  if (!hasProgramBody) {
    return {
      ok: false,
      errors: [
        'gnot source must include at least one executable clause or deploy/stack contract',
      ],
    };
  }

  const appId = headerMatch[1] ?? 'unknown-app';
  const version = headerMatch[2] ?? '0';

  return {
    ok: true,
    normalizedSource: normalized,
    manifest: {
      appId,
      version,
      hasDeployBlock: lines.some((line) => line.startsWith('deploy ')),
      declaredNamespaces: parseNamespaces(lines),
    },
  };
}

/**
 * An Observer geometrically formulates a brand new GG program physically in the void.
 * This binds the source code onto a compute topology that can be transferred or run.
 */
export function forgeTopology(
  creator: ObserverState,
  programName: string,
  sourceCode: string,
  compiledAST: TopologyASTNode[],
  options: ForgeTopologyOptions = {}
): ComputeTopology {
  const sourceLanguage = options.sourceLanguage ?? 'gg';
  const canonicalSource = normalizeSource(sourceCode);
  let gnotManifest: GnotTopologyManifest | undefined;

  if (sourceLanguage === 'gnot') {
    const validator = options.gnotValidator ?? validateGnotTopologySource;
    const validation = validator(canonicalSource);
    if (validation.ok === false) {
      throw new Error(
        `invalid_gnot_source: ${validation.errors.join('; ')}`
      );
    }
    gnotManifest = validation.manifest;
  }

  const hash = Math.random().toString(36).substring(2, 8);
  const normalizedName = programName.toLowerCase().replace(/[^a-z0-9]/g, '-');

  const digest = digestSource(canonicalSource);

  return {
    id: `aeon://topology/${normalizedName}-${hash}`,
    programName,
    authorDid: creator.did,
    ownerDid: creator.did,
    sourceLanguage,
    sourceCode: canonicalSource,
    sourceGGCode: canonicalSource,
    sourceDigest: digest,
    gnotManifest,
    ast: compiledAST,
    currentTickIndex: 0,
  };
}

/**
 * Forge a topology from first-class gnot source.
 */
export function forgeGnotTopology(
  creator: ObserverState,
  programName: string,
  sourceGnotCode: string,
  compiledAST: TopologyASTNode[],
  options: Omit<ForgeTopologyOptions, 'sourceLanguage'> = {}
): ComputeTopology {
  return forgeTopology(creator, programName, sourceGnotCode, compiledAST, {
    ...options,
    sourceLanguage: 'gnot',
  });
}

/**
 * Trade a compute topology between observers.
 * This alters the owner Did, transferring physical ownership of the algorithm.
 */
export function tradeTopology(
  topology: ComputeTopology,
  currentOwner: ObserverState,
  newOwner: ObserverState
): ComputeTopology {
  if (topology.ownerDid !== currentOwner.did) {
    throw new Error('You cannot transfer a topology you do not physically possess.');
  }

  // The compute mesh is transferred
  return {
    ...topology,
    ownerDid: newOwner.did
  };
}

function calculateTickEnergyCost(observer: ObserverState): number {
  const beta1 = Math.abs(DIMENSION_TIME_DILATION[observer.highestDimension]) || 1;
  const phyleDiscount = observer.majorPhyle === 'scientist' ? 0.8 : 1.0;
  return Math.max(1, Math.floor((10 * beta1) * phyleDiscount));
}

export interface TopologyExecutionResult {
  readonly state: ObserverState;
  readonly nextTopologyState: ComputeTopology;
  readonly executedNodes: number;
  readonly haltedDueToEnergy: boolean;
}

/**
 * Execute exactly N ticks of a GG Topology bound by physical thermodynamic limits.
 */
export function evaluateTopologyTicks(
  observer: ObserverState,
  topology: ComputeTopology,
  requestedTicks: number = 1
): TopologyExecutionResult {
  if (topology.ownerDid !== observer.did) {
    throw new Error('You cannot evaluate a topology you do not possess.');
  }

  const costPerTick = calculateTickEnergyCost(observer);

  let currentEnergy = observer.energy;
  let executedNodes = 0;
  let halted = false;
  let newTickIndex = topology.currentTickIndex;

  for (let i = 0; i < requestedTicks; i++) {
    if (newTickIndex >= topology.ast.length) {
      break;
    }
    if (currentEnergy < costPerTick) {
      halted = true;
      break;
    }
    currentEnergy -= costPerTick;
    executedNodes++;
    newTickIndex++;
  }

  const newState: ObserverState = { ...observer, energy: currentEnergy };
  const nextTopology: ComputeTopology = { ...topology, currentTickIndex: newTickIndex };

  return { state: newState, nextTopologyState: nextTopology, executedNodes, haltedDueToEnergy: halted };
}
