// @a0n/gnosis/src/void-quests.ts
/**
 * The Void — Peaceful Quests (The Meaning of Life)
 *
 * "If you can beat the system without conflict, you are a god."
 * 
 * Quests in the Gnosis Engine are not about conflict. There are no "Kill X Knots" tasks.
 * Quests are the pursuit of The Higher Arts, Culture, Ethos, Community, and Diversity.
 * 
 * The universe mathematics spawn Cultural Attractors where diversity is low or 
 * community cohesion has dropped. Traveling to these attractors and performing
 * Non-Quantifiable Excellence (Affirmation, Entanglement) stabilizes the fabric
 * of reality, generating massive Wisdom payouts.
 *
 * @module @a0n/gnosis/void-quests
 */

import type { GameCoordinate, ObserverState, DimensionIndex } from './void-game-types';

export type CulturalIdeal = 
  | 'community_cohesion' 
  | 'cultural_excellence' 
  | 'diversity_affirmation' 
  | 'higher_arts_pursuit'
  | 'peace_manifestation';

export interface MeaningOfLifeQuest {
  readonly questId: string;
  readonly targetCoordinate: GameCoordinate;
  readonly ideal: CulturalIdeal;
  /** Number of unique Phyles (diversity) required to stabilize the coordinate simultaneously */
  readonly requiredDiversity: number;
  /** Wisdom granted upon successful peaceful resolution */
  readonly wisdomYield: number;
  readonly isResolved: boolean;
}

/**
 * The void spontaneously demands cultural stabilization in certain shards.
 */
export function spawnCulturalAttractor(
  dimension: DimensionIndex,
  hexQ: number,
  hexR: number,
  hexS: number,
  ideal: CulturalIdeal
): MeaningOfLifeQuest {
  // Higher dimensional cultural goals require far more diversity (community building)
  const requiredDiversity = Math.min(7, 1 + Math.floor(dimension / 2));
  
  // The yield is astronomical compared to combat paths.
  // This explicitly rewards "beating the system without conflict".
  const wisdomYield = 100 * (dimension + 1) * requiredDiversity;

  return {
    questId: `quest-${ideal}-${dimension}-${Date.now()}`,
    targetCoordinate: { seed: 0, dimension, hexQ, hexR, hexS, tick: 0 },
    ideal,
    requiredDiversity,
    wisdomYield,
    isResolved: false
  };
}

export interface QuestResolutionResult {
  readonly states: ObserverState[];
  readonly resolvedQuest: MeaningOfLifeQuest;
}

/**
 * Resolve a Meaning of Life task.
 * This cannot be done randomly; it requires a community of highly diverse observers 
 * physically gathering on the exact spatial coordinate.
 */
export function resolvePeacefulQuest(
  participants: ObserverState[],
  quest: MeaningOfLifeQuest
): QuestResolutionResult {
  if (quest.isResolved) {
    throw new Error('This cultural ideal has already been fully manifested.');
  }

  // 1. Verify Locality
  for (const obs of participants) {
    if (
      obs.coordinate.dimension !== quest.targetCoordinate.dimension ||
      obs.coordinate.hexQ !== quest.targetCoordinate.hexQ ||
      obs.coordinate.hexR !== quest.targetCoordinate.hexR ||
      obs.coordinate.hexS !== quest.targetCoordinate.hexS
    ) {
      throw new Error(`Observer ${obs.did} is not physically present to build community.`);
    }
  }

  // 2. Verify Diversity ("the things that are good for us that we are bad at")
  const uniquePhyles = new Set(participants.map(p => p.majorPhyle));
  if (uniquePhyles.size < quest.requiredDiversity) {
    throw new Error(
      `Cultural stabilization requires a diversity score of ${quest.requiredDiversity}. ` +
      `Current community diversity is only ${uniquePhyles.size}. Bring more diverse Phyles.`
    );
  }

  // 3. Verify Lack of Conflict
  // A true mathematical Angel achieves this without sinning. But we accept any community.
  // The wisdom payout is MULTIPLIED if the participants have 0 sin (Absolute Pacifism).
  const isFlawlessPacifism = participants.every(p => p.sin === 0);
  const finalWisdomYield = isFlawlessPacifism ? quest.wisdomYield * 2 : quest.wisdomYield;

  const resolvedStates = participants.map(obs => ({
    ...obs,
    wisdom: obs.wisdom + finalWisdomYield,
    totalSuccesses: obs.totalSuccesses + 1
  }));

  return {
    states: resolvedStates,
    resolvedQuest: { ...quest, isResolved: true }
  };
}
