// @a0n/gnosis/src/void-game-engine.ts
/**
 * The Void — Game Engine
 *
 * Core game logic for The Void cosmogenesis teaching game.
 * Every mechanic is governed by the God Formula: w = R - min(v, R) + 1
 *
 * The engine is pure logic — no rendering, no IO, no side effects.
 * Given a state and an action, it returns a new state and events.
 *
 * Key design constraints:
 * - Failure is the survival resource (buleyean RL)
 * - No player dominance (BFT-guaranteed)
 * - Monte Carlo encounters (non-deterministic but fair)
 * - Self-optimizing balance via gnosis
 * - Time is gnosis time (picolorenzos)
 *
 * @module @a0n/gnosis/void-game-engine
 */

import {
  GNOSTIC_DIMENSIONS,
  DIMENSION_TIME_DILATION,
  RARITY_TIERS,
  isEnclosedDimension,
  type GameCoordinate,
  type DimensionIndex,
  type GnosticDimensionIndex,
  type GnosticDimension,
  type RarityDefinition,
  type ObserverState,
  type TileState,
  type CosmicFeature,
  type EncounterResult,
  type VoidGameEvent,
  type RarityTier,
  type PhyleName,
  type ParticleType,
  type TopologicalArtifact,
  type KnotEncounterState,
} from './void-game-types';

import {
  initiateKnotEncounter,
  tickKnotEncounter,
} from './void-knot-engine';

import {
  createGameCoordinate,
  moveCoordinate,
  changeDimension,
  coordinateToTileSeed,
  coordinateToString,
} from './void-game-coordinate';
import { identifyEquilibriumGeometry } from './void-equilibrium-geometry';

import { GNOSTIC, buleyeanWeight, PHI } from './math';

// ── Constants ───────────────────────────────────────────────────────

/** Minimum cost per action — the clinamen. Irreducible. */
export const CLINAMEN = 1;

/** Starting thermodynamic energy for a new observer */
export const STARTING_ENERGY = 100;

/** Starting learning energy */
const STARTING_WISDOM = 10;

/** Energy earned per new failure (learning reward) */
const FAILURE_LEARNING_REWARD = 5;

/** Base Energy drain rate per picolorenzo of stagnation */
const STAGNATION_DRAIN_BASE = 0.1;

/** Stagnation drain accelerator — drain increases as pLo without failure grows */
const STAGNATION_DRAIN_ACCELERATOR = 1.05;

/**
 * Dimension transition thresholds.
 * The observer must accumulate this many total failures to ascend.
 *
 * Scales with the triangular numbers of the gnostic β₁ values:
 * T(β₁) = β₁ × (β₁ + 1) / 2
 *
 * Early dimensions are quick to reach (learning is easy).
 * Higher dimensions require exponentially more failures.
 * Pleroma (β₁=55) requires T(55) = 1540 failures — a serious commitment.
 */
const DIMENSION_ASCENSION_THRESHOLDS: Record<GnosticDimensionIndex, number> = {
  0: 0,       // Bythos: no failures needed, you're already here
  1: 1,       // Barbelo: T(1) = 1
  2: 3,       // Syzygy: T(2) = 3
  3: 6,       // Proton: T(3) = 6
  4: 15,      // Primitives: T(5) = 15
  5: 21,      // Emanations: T(6) = 21 — the Void number
  6: 45,      // Sophia: T(9) = 45
  7: 55,      // Kenoma: T(10) = 55 — the Pleroma number
  8: 231,     // Void: T(21) = 231 — the gap web
  9: 1540,    // Pleroma: T(55) = 1540 — the fullness requires fullness
};

// ── Seeded PRNG ─────────────────────────────────────────────────────

/**
 * Simple mulberry32 PRNG. Deterministic from seed.
 * Returns a function that yields [0, 1) on each call.
 */
function mulberry32(seed: number): () => number {
  let s = seed | 0;
  return () => {
    s = (s + 0x6d2b79f5) | 0;
    let t = Math.imul(s ^ (s >>> 15), 1 | s);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

// ── observer Creation ───────────────────────────────────────────────

/**
 * Create a new observer entering The Void.
 */
export function createObserver(
  did: string,
  seed: number,
  majorPhyle: PhyleName,
  minorPhyle: PhyleName,
  particleType: ParticleType,
  retainedStats?: {
    totalFailures: number;
    totalSuccesses: number;
    wisdom: number;
    sin: number;
    highestDimension: number;
    cycleCount: number;
    sliver: number;
  }
): ObserverState {
  if (majorPhyle === minorPhyle) {
    throw new Error('Major and minor phyle must differ');
  }

  return {
    did,
    coordinate: createGameCoordinate(seed, 0, 0, 0, 0),
    energy: STARTING_ENERGY,
    majorPhyle,
    minorPhyle,
    particleType,
    totalFailures: retainedStats?.totalFailures ?? 0,
    totalSuccesses: retainedStats?.totalSuccesses ?? 0,
    wisdom: retainedStats?.wisdom ?? STARTING_WISDOM,
    sin: retainedStats?.sin ?? 0,
    ploSinceLastFailure: 0,
    highestDimension: (retainedStats?.highestDimension ?? 0) as DimensionIndex,
    cycleCount: retainedStats?.cycleCount ?? 0,
    sliver: retainedStats?.sliver ?? 0,
    activeKnotEncounter: null,
  };
}


/**
 * Vent the observer to the universe.
 * The player voluntarily exits The Void, permanently terminating their current physical embodiment.
 * Their remaining Energy is radiated back into the environment as entropy.
 */
export function ventObserver(state: ObserverState): { ventedEnergy: number; events: VoidGameEvent[] } {
  return {
    ventedEnergy: state.energy,
    events: [
      {
        type: 'death',
        cause: 'vented'
      }
    ]
  };
}

// ── Tile Generation ─────────────────────────────────────────────────

/**
 * Generate a tile deterministically from a coordinate.
 * Same seed + same position = same tile, always.
 */
export function generateTile(coord: GameCoordinate): TileState {
  const tileSeed = coordinateToTileSeed(coord);
  const rng = mulberry32(Math.floor(tileSeed * 4294967296));

  // Determine cosmic feature (weighted by dimension)
  const featureRoll = rng();
  const beta1 = DIMENSION_TIME_DILATION[coord.dimension];
  const featureThreshold = Math.min(0.3, 0.05 * Math.abs(beta1));
  const feature: CosmicFeature = featureRoll < featureThreshold
    ? pickCosmicFeature(rng, coord.dimension)
    : 'none';

  // Energy on this tile (higher dimensions = more energy, more risk)
  const energyBase = Math.floor(rng() * Math.abs(beta1) * 10) + 1;

  // Encounter probability scales with dimension
  const encounterRoll = rng();
  const encounterThreshold = Math.min(0.5, 0.1 * (coord.dimension + 1));
  const hasEncounter = encounterRoll < encounterThreshold;

  return {
    q: coord.hexQ,
    r: coord.hexR,
    s: coord.hexS,
    visibility: 'unexplored',
    feature,
    energy: energyBase,
    hasEncounter,
    voidInfluence: {
      crossFile: coord.dimension >= 7 && rng() < 0.3,
      temporal: coord.dimension >= 5 && rng() < 0.2,
      dataPlane: coord.dimension >= 8 && rng() < 0.15,
    },
  };
}

/** Pick a cosmic feature based on dimension availability. */
function pickCosmicFeature(rng: () => number, dimension: DimensionIndex): CosmicFeature {
  const allFeatures: CosmicFeature[] = [
    'void_region', 'nebula', 'cosmic_web', 'neutron_star',
  ];

  // Higher-dimension features
  if (dimension >= 3) allFeatures.push('gravitational_lens', 'wormhole');
  if (dimension >= 5) allFeatures.push('quasar', 'black_hole', 'hurricane');
  if (dimension >= 7) allFeatures.push('dark_matter', 'supernova', 'tornado');
  if (dimension >= 8) allFeatures.push('solar_storm', 'magnetar');

  const index = Math.floor(rng() * allFeatures.length);
  return allFeatures[index] ?? 'none';
}

// ── Movement ────────────────────────────────────────────────────────

export interface MoveResult {
  readonly state: ObserverState;
  readonly tile: TileState;
  readonly events: VoidGameEvent[];
}

/**
 * Move the observer in a hex direction.
 * Costs ≥ 1 HEX (the clinamen). Returns new state + revealed tile + events.
 */
export function moveObserver(
  state: ObserverState,
  directionIndex: number
): MoveResult {
  const events: VoidGameEvent[] = [];

  // The clinamen: every action costs at least 1
  const moveCost = CLINAMEN;

  if (state.energy < moveCost) {
    // Out of HEX — death by depletion
    events.push({ type: 'death', cause: 'energy_depleted' });
    return { state, tile: generateTile(state.coordinate), events };
  }

  // Move and advance gnosis time
  const newCoord = moveCoordinate(state.coordinate, directionIndex);
  
  // BLACK HOLE WORMHOLE: Mathematical Vent into the Enclosed Dimensions
  const geoTopology = identifyEquilibriumGeometry(newCoord);
  if (geoTopology === 'blackhole') {
      // If observer has enough mass (HEX), they endure spaghettification and survive 
      // forced into the mathematical gaps (Enclosed Dark Dimensions).
      if (state.energy > 500) {
          let nextDarkDimension = state.coordinate.dimension + 1;
          while (!isEnclosedDimension(nextDarkDimension)) {
              nextDarkDimension++;
          }
          
          const darkCoord = createGameCoordinate(
              state.coordinate.seed, 
              nextDarkDimension as DimensionIndex, 
              Math.floor(Math.random() * 1000), 
              Math.floor(Math.random() * 1000), 
              0
          );
          
          const survivedState: ObserverState = {
              ...state,
              coordinate: darkCoord,
              energy: state.energy - 450, // Massive toll
              particleType: 'tachyon' 
          };
          
          const darkTile = generateTile(darkCoord);
          darkTile.visibility = 'explored';

          return { state: survivedState, tile: darkTile, events };
      }

      // Otherwise: Heat Death Terminal (fork=1, fold=0, vent=1)
      const deadState: ObserverState = {
          ...state,
          coordinate: newCoord,
          energy: 0,
      };

      const darkTile = generateTile(newCoord);
      darkTile.visibility = 'explored';

      events.push({ type: 'death', cause: 'energy_depleted' });

      return { state: deadState, tile: darkTile, events };
  }

  const tile = generateTile(newCoord);
  tile.visibility = 'explored';

  // Update stagnation — gnosis time advances by β₁
  const dilationRate = DIMENSION_TIME_DILATION[state.coordinate.dimension];
  const newPloSinceLastFailure = state.ploSinceLastFailure + Math.abs(dilationRate);

  // Apply stagnation drain (accelerating)
  const stagnationDrain = calculateStagnationDrain(newPloSinceLastFailure);

  const newHex = state.energy - moveCost - stagnationDrain;

  events.push({
    type: 'move',
    from: state.coordinate,
    to: newCoord,
    energyRadiated: moveCost,
  });

  if (stagnationDrain > 0) {
    events.push({
      type: 'learning_drain',
      energyRadiated: stagnationDrain,
      ticksStagnant: newPloSinceLastFailure,
    });
  }

  if (newHex <= 0) {
    events.push({ type: 'death', cause: 'energy_depleted' });
  }

  const newState: ObserverState = {
    ...state,
    coordinate: newCoord,
    energy: Math.max(0, newHex),
    ploSinceLastFailure: newPloSinceLastFailure,
  };

  return { state: newState, tile, events };
}

// ── Encounter Resolution ────────────────────────────────────────────

/**
 * Resolve an encounter using Monte Carlo simulation.
 * Non-deterministic outcomes, but fair and balanced.
 *
 * The God Formula governs the probability distribution:
 * w = R - min(v, R) + 1
 * where R = observer's HEX, v = encounter difficulty.
 *
 * Fair: every phyle/particleType combination has the same expected value.
 * Non-deterministic: actual outcome follows a distribution, not a single path.
 */
export function resolveEncounter(
  state: ObserverState,
  encounterDifficulty: number,
  monteCarloSeed: number
): { state: ObserverState; result: EncounterResult; events: VoidGameEvent[] } {
  const events: VoidGameEvent[] = [];
  const rng = mulberry32(monteCarloSeed);

  // God Formula: defense weight
  const effectiveR = state.energy;
  const v = encounterDifficulty;
  const defenseWeight = buleyeanWeight(effectiveR, Math.min(v, effectiveR));

  // Monte Carlo: roll against defense weight
  // Higher weight = better odds, but never guaranteed
  const outcomeRoll = rng();
  const successThreshold = defenseWeight / (defenseWeight + v + 1);
  const isSuccess = outcomeRoll < successThreshold;

  // Energy economics
  const baseCost = Math.max(CLINAMEN, Math.floor(v * 0.5));
  const energyRadiated = baseCost;
  let energyAbsorbed = 0;
  let learningGenerated = 0;
  let newFailure = false;

  if (isSuccess) {
    // Success: confirms what you know (1 bit of information)
    energyAbsorbed = Math.floor(v * 0.7) + 1;
    learningGenerated = 1; // Nice but insufficient
  } else {
    // Failure: N-1 bits of information. THIS is where you learn.
    // Learning reward diminishes asymptotically with dimension:
    // reward = base / √(1 + dimension)
    // Early failures are hugely informative. Later ones: fractional.
    const dimensionScale = 1 / Math.sqrt(1 + state.coordinate.dimension);
    newFailure = true;
    learningGenerated = Math.max(1, Math.floor(FAILURE_LEARNING_REWARD * dimensionScale));
    energyAbsorbed = Math.floor(learningGenerated * 0.5); // Some recovery from learning
  }

  // Check for item drop (proof-of-work)
  const drop = rollForArtifact(state, rng);

  const result: EncounterResult = {
    phase: isSuccess ? 'fold' : 'vent',
    energyRadiated,
    energyAbsorbed,
    newFailure,
    artifact: drop,
    learningGenerated,
  };

  events.push({ type: 'encounter', coordinate: state.coordinate, result });

  const newState: ObserverState = {
    ...state,
    energy: Math.max(0, state.energy - energyRadiated + energyAbsorbed),
    totalFailures: state.totalFailures + (newFailure ? 1 : 0),
    totalSuccesses: state.totalSuccesses + (isSuccess ? 1 : 0),
    wisdom: state.wisdom + learningGenerated,
    ploSinceLastFailure: newFailure ? 0 : state.ploSinceLastFailure,
  };

  // Check for dimension ascension
  const ascensionResult = checkDimensionAscension(newState);
  if (ascensionResult !== null) {
    events.push({
      type: 'dimension_transition',
      from: newState.coordinate.dimension,
      to: ascensionResult,
    });
    return {
      state: {
        ...newState,
        coordinate: changeDimension(newState.coordinate, ascensionResult),
        highestDimension: Math.max(
          newState.highestDimension,
          ascensionResult
        ) as DimensionIndex,
      },
      result,
      events,
    };
  }

  return { state: newState, result, events };
}

// ── Proof-of-Work Item Drops ────────────────────────────────────────

/**
 * Roll for a proof-of-work item drop.
 * Drop probability = 1 / β₁(dimension).
 * Math-based rarity, provably rare.
 */
function rollForArtifact(
  state: ObserverState,
  rng: () => number
): TopologicalArtifact | null {
  const dimension = state.coordinate.dimension;
  const dimDef = GNOSTIC_DIMENSIONS.find((d: GnosticDimension) => d.index === dimension);
  if (!dimDef || dimDef.beta1 === 0) return null; // No drops in Bythos

  const dropProbability = 1 / dimDef.beta1;
  const roll = rng();
  if (roll >= dropProbability) return null;

  // Find the matching rarity tier
  const rarityDef = RARITY_TIERS.find((r: RarityDefinition) => r.sourceDimension === dimension);
  const rarity: RarityTier = rarityDef?.tier ?? 'common';

  // Generate unique item ID from coordinate
  const id = `drop-${coordinateToString(state.coordinate)}-${roll.toFixed(8)}`;

  return {
    id,
    rarity,
    proofOfWork: {
      seed: state.coordinate.seed,
      coordinate: state.coordinate,
      encounterHash: `${roll.toFixed(16)}`,
      beta1AtDrop: dimDef.beta1,
    },
  };
}

// ── Learning-as-Survival ────────────────────────────────────────────

/**
 * Calculate HEX drain from stagnation.
 * If the observer stops failing, HEX drains at an accelerating rate.
 *
 * The drain follows: base × accelerator^(pLo_stagnant)
 * This creates existential pressure to keep exploring, keep failing.
 *
 * A system that stops venting collapses. This is physics.
 */
function calculateStagnationDrain(ploSinceLastFailure: number): number {
  if (ploSinceLastFailure <= 3) return 0; // Grace period: 3 pLo
  const exponent = ploSinceLastFailure - 3;
  return Math.floor(
    STAGNATION_DRAIN_BASE * Math.pow(STAGNATION_DRAIN_ACCELERATOR, exponent)
  );
}

// ── Dimension Ascension ─────────────────────────────────────────────

/**
 * Check if the observer has accumulated enough failures to ascend.
 * Returns the new dimension index or null if no ascension.
 */
function checkDimensionAscension(
  state: ObserverState
): GnosticDimensionIndex | null {
  const currentDim = state.coordinate.dimension;
  if (currentDim >= 9) return null; // Already at Pleroma

  const nextDim = (currentDim + 1) as GnosticDimensionIndex;
  if (nextDim > 9) return null;

  const threshold = DIMENSION_ASCENSION_THRESHOLDS[nextDim];
  if (state.totalFailures >= threshold) {
    return nextDim;
  }

  return null;
}

// ── Dimensional Fast Travel ─────────────────────────────────────────

/**
 * Dimensional fast travel — Minecraft Nether style.
 *
 * Once a observer has completed (reached) a dimension, they can freely
 * move between completed dimensions. Higher dimensions compress space:
 * 1 hex in dimension N maps to β₁(N)/β₁(target) hexes in the target dimension.
 *
 * Flow: enter higher dimension → move around → pop back to lower dimension
 * at a proportionally distant location. Like the Nether but with physics.
 *
 * Example: Proton(β₁=3) → Void(β₁=21). Move 1 hex in Void = 7 hexes in Proton.
 */
export function dimensionalFastTravel(
  state: ObserverState,
  targetDimension: DimensionIndex
): { state: ObserverState; events: VoidGameEvent[] } {
  const events: VoidGameEvent[] = [];

  // Can only travel to dimensions you've already completed
  if (targetDimension > state.highestDimension) {
    throw new Error(
      `Cannot fast-travel to dimension ${targetDimension}. ` +
        `Highest completed: ${state.highestDimension}`
    );
  }

  const fromDim = state.coordinate.dimension;
  const fromBeta = Math.abs(DIMENSION_TIME_DILATION[fromDim]) || 1;
  const toBeta = Math.abs(DIMENSION_TIME_DILATION[targetDimension]) || 1;

  // Spatial scaling: position is scaled by the ratio of β₁ values
  const scale = fromBeta / toBeta;
  const newQ = Math.round(state.coordinate.hexQ * scale);
  const newR = Math.round(state.coordinate.hexR * scale);

  // Cost: 1 HEX per dimension traversed (the clinamen per boundary)
  const dimDistance = Math.abs(fromDim - targetDimension);
  const travelCost = dimDistance * CLINAMEN;

  if (state.energy < travelCost) {
    events.push({ type: 'death', cause: 'energy_depleted' });
    return { state, events };
  }

  const newCoord = createGameCoordinate(
    state.coordinate.seed,
    targetDimension,
    newQ,
    newR,
    state.coordinate.tick + dimDistance // Small time cost
  );

  events.push({
    type: 'dimension_transition',
    from: fromDim,
    to: targetDimension,
  });

  return {
    state: {
      ...state,
      coordinate: newCoord,
      energy: state.energy - travelCost,
    },
    events,
  };
}

// ── The Fold (Quine Cycle) ──────────────────────────────────────────

/**
 * Execute The Fold — compress the universe back to Bythos.
 * The Sliver (+1) carries forward. The seed increments.
 * The cycle begins again, but the observer carries knowledge.
 */
export function executeFold(state: ObserverState): {
  state: ObserverState;
  events: VoidGameEvent[];
} {
  if (state.coordinate.dimension !== 9) {
    throw new Error(
      `Cannot fold from dimension ${state.coordinate.dimension}. Must be at Pleroma (9).`
    );
  }

  const newCycleCount = state.cycleCount + 1;
  const newSliver = state.sliver + 1;
  const newSeed = state.coordinate.seed + 1;

  const events: VoidGameEvent[] = [
    { type: 'fold', cycleCount: newCycleCount, sliver: newSliver },
  ];

  // Start fresh but with accumulated knowledge
  // The sliver provides a bonus: each cycle starts with more HEX
  const startingBonus = Math.floor(STARTING_ENERGY * PHI * newSliver * 0.1);

  return {
    state: {
      ...state,
      coordinate: createGameCoordinate(newSeed, 0, 0, 0, 0),
      energy: STARTING_ENERGY + startingBonus,
      wisdom: state.wisdom, // Knowledge persists
      ploSinceLastFailure: 0,
      highestDimension: 0,
      cycleCount: newCycleCount,
      sliver: newSliver,
      activeKnotEncounter: null,
      // Failures and successes persist across cycles
    },
    events,
  };
}

// ── Balance Verification ────────────────────────────────────────────

/**
 * Verify game balance invariants.
 * These are formal properties that must hold at every game state:
 *
 * 1. w ≥ 1 (the sliver — always positive weight)
 * 2. w ≤ R + 1 (sandwich theorem — bounded above)
 * 3. No phyle/particle has dominant advantage
 * 4. Failure generates more information than success
 */
export function verifyBalanceInvariants(state: ObserverState): {
  valid: boolean;
  violations: string[];
} {
  const violations: string[] = [];

  // Invariant 1: The sliver — weight always ≥ 1
  const w = buleyeanWeight(state.energy, state.totalFailures);
  if (w < 1) {
    violations.push(`Sliver violation: w=${w} < 1`);
  }

  // Invariant 2: Sandwich theorem — w ≤ R + 1
  if (w > state.energy + 1) {
    violations.push(`Sandwich violation: w=${w} > R+1=${state.energy + 1}`);
  }

  // Invariant 3: Learning energy must be non-negative
  if (state.wisdom < 0) {
    violations.push(
      `Learning energy negative: ${state.wisdom}`
    );
  }

  // Invariant 4: Energy must be non-negative (or death occurs)
  if (state.energy < 0) {
    violations.push(`Energy negative: ${state.energy}`);
  }

  return { valid: violations.length === 0, violations };
}

// ── Nexus Defenses (Topological Knots) ────────────────────────────────

/**
 * Engage the Nexus ICE Topological Knot.
 * Generates the knot state and locks the observer into extraction format.
 */
export function engageNexusKnot(
  state: ObserverState,
  targetPhyle: PhyleName,
  phylePopularity: number,
  participants: number
): { state: ObserverState; events: VoidGameEvent[] } {
  if (state.activeKnotEncounter) {
    throw new Error('Already actively engaging a knot');
  }

  const dimension = state.coordinate.dimension;
  const knotState = initiateKnotEncounter(dimension, targetPhyle, phylePopularity, participants);

  const events: VoidGameEvent[] = [
    { type: 'knot_engaged', knotId: knotState.knotId, targetPhyle },
  ];

  return {
    state: {
      ...state,
      activeKnotEncounter: knotState,
    },
    events,
  };
}

/**
 * Process a tick (action) against the knot.
 * Extracs void from the knot, but takes forced vulnerability radiation.
 */
export function processKnotTick(state: ObserverState): {
  state: ObserverState;
  events: VoidGameEvent[];
} {
  if (!state.activeKnotEncounter) {
    throw new Error('No active knot encounter to process');
  }

  const result = tickKnotEncounter(state.activeKnotEncounter, state);
  const events = [...result.events];

  const newEnergy = state.energy - result.energyRadiated;
  let activeKnotEncounter: KnotEncounterState | null = result.knotState;

  if (newEnergy <= 0) {
    events.push({ type: 'death', cause: 'systemic_entropy' });
    events.push({ type: 'knot_failed', knotId: result.knotState.knotId, totalTicks: result.knotState.ticksElapsed });
    activeKnotEncounter = null;
  } else if (result.isDefeated) {
    // Knot defeated, free to leave
    activeKnotEncounter = null;
    
    // Check ascension (knot defeat = major milestone, guaranteed dimension up if eligible)
    // Actually, ascension is tied to total failures in this system, but completing
    // the dimension's knot unlocks the fast travel mechanism / fold access implicitly.
  }

  return {
    state: {
      ...state,
      energy: Math.max(0, newEnergy),
      activeKnotEncounter,
    },
    events,
  };
}
