// @a0n/gnosis/src/wallington-rotation.ts
/**
 * Wallington Rotation — Pipelined Multi-Stage Scheduling Topology
 *
 * Maps the Wallington Rotation topology directly to scheduling geometry:
 *   - chunks = independent parallel streams / paths
 *   - stages = sequential pipeline stages
 *   - β₁ = chunks - 1 = independent routing paths
 *
 * It tracks optimality limits as prescribed in WallingtonOptimality.lean,
 * maximizing makespan efficiency by parallelizing strictly independent chunks.
 *
 * @packageDocumentation
 */

import { GnosticMeshLevel, GNOSTIC_MESHES } from './spiderweb-topology';

export interface WallingtonConfig {
  /** Number of independent parallel data chunks */
  readonly chunks: number;
  /** Number of sequential pipeline stages */
  readonly stages: number;
}

export type WallingtonRegime = 'sequential' | 'pipeline' | 'saturated';

export interface WallingtonTopologyReport {
  readonly config: WallingtonConfig;
  readonly routingPaths: number;
  readonly speedupEstimate: number;
  readonly regime: WallingtonRegime;
}

export interface WallingtonDeficitReport {
  readonly required: number;
  readonly available: number;
  readonly deficit: number;
  readonly utilization: number;
  readonly assessment: string;
}

// ─── Core Computations ──────────────────────────────────────────────

/** β₁ = c - 1: independent parallel routing paths (cycles). */
export function wallingtonRoutingPaths(config: WallingtonConfig): number {
  return Math.max(0, config.chunks - 1);
}

/** The makespan bound correlation directly maps to speedup factor (from Lean proofs) */
export function wallingtonSpeedupFactor(config: WallingtonConfig): number {
  return config.chunks;
}

/** Classify scheduling regime based on parallel chunks vs stages. */
export function classifyWallingtonRegime(config: WallingtonConfig): WallingtonRegime {
  if (config.chunks <= 1) return 'sequential';
  if (config.chunks > config.stages * 4) return 'saturated'; // Arbitrary saturation definition based on pipeline depth mapping
  return 'pipeline';
}

/** Routing deficit: required − available, floored at 0. */
export function wallingtonDeficit(required: number, config: WallingtonConfig): number {
  return Math.max(0, required - wallingtonRoutingPaths(config));
}

// ─── Predictive Dimensional Metric ────────────────────────────────────

/** Lower Bound: Floor conditions for minimal operation.
 *  A schedule is sequential and lacks parallel rotation if chunks <= 1.
 */
export function requiresWallingtonLowerBound(chunks: number, _stages: number): boolean {
  return chunks <= 1;
}

/** Gain Threshold: When pipeline rotation saturation begins providing speedup.
 *  Requires at least 2 independent chunks and 1 stage to form a meaningful rotation.
 */
export function yieldsWallingtonGainThreshold(chunks: number, stages: number): boolean {
  return chunks >= 2 && stages >= 1;
}

/** Ceiling: Makespan critical path saturation.
 *  Saturation limit when coordination overhead equals sequential execution time.
 */
export function requiresWallingtonCeiling(chunks: number, stages: number): boolean {
  // Model empirical limit where high chunks outpace stages significantly
  return chunks >= 11 && stages >= 5; // Pleroma ceiling values
}

/**
 * Predicts the optimal Wallington scale based on workload concurrency and pipeline depth.
 * Maps back to the 2D Spiderweb Projection.
 */
export function predictWallingtonDimension(
  concurrency: number,
  depth: number,
  memoryConstraints: boolean = false
): GnosticMeshLevel {
  if (memoryConstraints) {
    return 'bythos'; // Fallback to sequential to save memory
  } else if (requiresWallingtonCeiling(concurrency, depth)) {
    return 'pleroma';
  } else if (concurrency >= 7 && depth >= 3) {
    return 'void';
  } else if (concurrency >= 5 && depth >= 2) {
    return 'kenoma';
  } else if (concurrency >= 3 && depth >= 3) {
    return 'sophia';
  } else if (concurrency >= 3 && depth >= 2) {
    return 'emanations';
  } else if (concurrency >= 5 && depth >= 1) {
    return 'primitives';
  } else if (concurrency >= 3 && depth >= 1) {
    return 'proton';
  } else if (yieldsWallingtonGainThreshold(concurrency, depth)) {
    return 'syzygy';
  } else if (concurrency === 1 && depth === 1) {
    return 'barbelo';
  } else {
    return 'bythos';
  }
}

/**
 * Automagically resolves the appropriate Wallington scale (2D Spiderweb Projection)
 * based on the number of parameters the pipeline processes.
 * 
 * @param numParameters Number of parameters in the workload (e.g. 70e9)
 */
export function resolveWallingtonDimension(numParameters: number): GnosticMeshLevel {
  if (numParameters >= 50_000_000_000) return 'pleroma'; // 55D -> 11 chunks, 5 stages
  if (numParameters >= 10_000_000_000) return 'void';    // 21D ->  7 chunks, 3 stages
  if (numParameters >= 3_000_000_000) return 'kenoma';   // 10D ->  5 chunks, 2 stages
  if (numParameters >= 1_000_000_000) return 'sophia';   // 9D  ->  3 chunks, 3 stages
  if (numParameters >= 500_000_000) return 'emanations'; // 6D  ->  3 chunks, 2 stages
  if (numParameters >= 100_000_000) return 'primitives'; // 5D  ->  5 chunks, 1 stages
  if (numParameters >= 10_000_000) return 'proton';      // 3D  ->  3 chunks, 1 stages
  if (numParameters >= 1_000_000) return 'syzygy';       // 2D  ->  2 chunks, 1 stages
  if (numParameters >= 100_000) return 'barbelo';        // 1D  ->  1 chunks, 1 stages
  return 'bythos';                                       // 0D  ->  0 chunks, 0 stages
}

/**
 * Generates the GG topology source for a Dynamic Wallington Rotation primitive.
 * Automagically sets chunk slicing and stage depth based on the 2D Spiderweb projection.
 * 
 * @param dimension The mapped GnosticMeshLevel determining geometry.
 */
export function generateDynamicWallingtonRotationTopology(
  dimension: GnosticMeshLevel
): string {
  const meshConfig = GNOSTIC_MESHES[dimension];
  const chunks = Math.max(1, meshConfig.peers); // k: peers map to chunks
  const stages = Math.max(1, meshConfig.hops);  // s: hops map to stages
  
  const stageNames = ['embed', 'attend', 'ffn', 'project'].slice(0, stages);
  
  // Backfill generic names if stages exceed default vocabulary
  while (stageNames.length < stages) {
    stageNames.push(`stage${stageNames.length}`);
  }
  
  const stageNamesStr = `[${stageNames.join(',')}]`;
  
  if (chunks <= 1 || stages <= 1) {
    return [
      `// Wallington Rotation (Sequential Scale - Dimension: ${dimension})`,
      `(prompt: FlowFrame { role: 'prompt-sequence' })`,
      `(sequential_output: WallingtonRotation { stages: '${stages}', chunks: '1', stageNames: '${stageNamesStr}' })`,
      `(prompt)-[:PROCESS]->(sequential_output)`
    ].join('\n');
  }

  const lines: string[] = [];
  lines.push(`// Dynamic Wallington Rotation (Scale: ${dimension}, Chunks: ${chunks}, Stages: ${stages})`);
  lines.push(`// Projected directly from Spiderweb Topology (peers->chunks, hops->stages)`);
  lines.push(`(prompt: FlowFrame { role: 'prompt-sequence' })`);
  lines.push(`(rotated_output: WallingtonRotation { stages: '${stages}', chunks: '${chunks}', stageNames: '${stageNamesStr}' })`);
  lines.push(`(prompt)-[:PROCESS]->(rotated_output)`);
  
  return lines.join('\n');
}

// ─── Analysis ───────────────────────────────────────────────────────

/** Full topology report for a wallington configuration. */
export function analyzeWallington(config: WallingtonConfig): WallingtonTopologyReport {
  return {
    config,
    routingPaths: wallingtonRoutingPaths(config),
    speedupEstimate: wallingtonSpeedupFactor(config),
    regime: classifyWallingtonRegime(config),
  };
}

/** Deficit report for a rotation against a routing demand. */
export function analyzeWallingtonDeficit(
  requiredPaths: number,
  config: WallingtonConfig
): WallingtonDeficitReport {
  const available = wallingtonRoutingPaths(config);
  const deficit = Math.max(0, requiredPaths - available);
  const utilization =
    requiredPaths === 0 ? 1 : Math.min(1, available / requiredPaths);

  let assessment: string;
  if (deficit === 0) {
    assessment = `Topology-matched: 0 Bules — rotation provides ${available} independent paths`;
  } else if (utilization >= 0.8) {
    assessment = `Near-matched: ${deficit} Bules — ${(utilization * 100).toFixed(0)}% coverage`;
  } else if (available === 0) {
    assessment = `Sequential bottleneck: ${deficit} Bules — rotation is essentially sequential`;
  } else {
    assessment = `Underprovisioned: ${deficit} Bules — rotation provides ${available} of ${requiredPaths} required paths`;
  }

  return { required: requiredPaths, available, deficit, utilization, assessment };
}
