// @a0n/gnosis/src/wasm-execution-bridge.ts
/// <reference lib="dom" />
// NOTE: `node:crypto`, `node:fs/promises`, and `node:path` are imported
// lazily inside the inline-bytes / file-path branches of `loadWasmModule` so
// the pre-compiled-Module path (moduleResolver / wasmModule) stays edge-safe:
// Cloudflare Workers consume deploy-time `CompiledWasm` Modules and must never
// reach runtime compile or filesystem code.
// `betty` is intentional — the Betty compiler namespace (`src/betty/compiler.ts`),
// distinct from `src/betti/` (which has no `compiler` module). Not a typo for "betti".
import type { ASTEdge, ASTNode, GraphAST } from './betty/compiler';
import { GnosisRegistry } from './runtime/registry';
import {
  GnosisEngine,
  type GnosisEngineExecutionResult,
  type GnosisEngineExecuteOptions,
  type GnosisEngineOptions,
} from './runtime/engine';

export type WasmExecutionAbi = 'scalar-v1' | 'json-v1';

export interface WasmNodeExecutionPlan {
  readonly node_id: string;
  readonly kind: 'entry' | 'call' | 'statement' | 'return' | 'join';
  readonly file_path?: string;
  readonly wasm_base64?: string;
  readonly wasm_module?: WebAssembly.Module;
  readonly export_name?: string;
  readonly abi?: WasmExecutionAbi;
  readonly memory_export?: string;
  readonly alloc_export?: string;
  readonly result_length_export?: string;
  readonly free_export?: string;
}

export interface WasmExecutionManifest {
  readonly artifact_id?: string;
  readonly file_path?: string;
  readonly wasm_base64?: string;
  /** Pre-compiled module (e.g. a Workers `CompiledWasm` import) — never recompiled from bytes. */
  readonly wasm_module?: WebAssembly.Module;
  readonly export_name?: string;
  readonly abi?: WasmExecutionAbi;
  readonly memory_export?: string;
  readonly alloc_export?: string;
  readonly result_length_export?: string;
  readonly free_export?: string;
  readonly resolve_relative_to?: string;
  readonly node_execution_plans: readonly WasmNodeExecutionPlan[];
}

export interface ResolvedWasmExecutionDescriptor {
  readonly artifactId?: string;
  readonly filePath?: string;
  readonly wasmBase64?: string;
  readonly wasmModule?: WebAssembly.Module;
  readonly exportName: string;
  readonly abi: WasmExecutionAbi;
  readonly memoryExport: string;
  readonly allocExport: string;
  readonly resultLengthExport: string;
  readonly freeExport?: string;
  readonly resolveRelativeTo?: string;
}

interface WasmBridgeState {
  readonly input: unknown;
  readonly locals: Record<string, unknown>;
  readonly descriptor: ResolvedWasmExecutionDescriptor;
}

export interface LoadedWasmModule {
  readonly module: WebAssembly.Module;
  readonly imports: readonly WebAssembly.ModuleImportDescriptor[];
}

/**
 * Resolve a descriptor to an already-compiled module (e.g. a deploy-time
 * `CompiledWasm` import on Cloudflare Workers). Returning `null` falls
 * through to the descriptor's own `wasmModule` / inline base64 / file path.
 */
export type WasmModuleResolver = (
  descriptor: Readonly<ResolvedWasmExecutionDescriptor>
) =>
  | Promise<LoadedWasmModule | WebAssembly.Module | null>
  | LoadedWasmModule
  | WebAssembly.Module
  | null;

export interface WasmExecutionHostOptions {
  readonly importObject?:
    | WebAssembly.Imports
    | ((
        descriptor: Readonly<ResolvedWasmExecutionDescriptor>
      ) => WebAssembly.Imports | Promise<WebAssembly.Imports>);
  readonly moduleResolver?: WasmModuleResolver;
}

export interface ExecuteWasmOptions extends WasmExecutionHostOptions {
  readonly ast: GraphAST;
  readonly manifest: WasmExecutionManifest;
  readonly input?: unknown;
  readonly registry?: GnosisRegistry;
  readonly engineOptions?: GnosisEngineOptions;
  readonly executeOptions?: GnosisEngineExecuteOptions;
}

const DEFAULT_WASM_EXPORT_NAME = 'main';
const DEFAULT_WASM_ABI: WasmExecutionAbi = 'json-v1';
const DEFAULT_MEMORY_EXPORT = 'memory';
const DEFAULT_ALLOC_EXPORT = 'gnosis_alloc';
const DEFAULT_RESULT_LENGTH_EXPORT = 'gnosis_result_len';
const DEFAULT_FREE_EXPORT = 'gnosis_free';
const DEFAULT_ASSIGN_TO = '__wasm_result__';

const wasmModuleCache = new Map<string, Promise<LoadedWasmModule>>();

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function resolveContextPath(value: unknown, path: string): unknown {
  const segments = path
    .split('.')
    .map((segment) => segment.trim())
    .filter((segment) => segment.length > 0);

  let current: unknown = value;
  for (const segment of segments) {
    if (Array.isArray(current)) {
      const index = Number.parseInt(segment, 10);
      if (!Number.isInteger(index) || index < 0 || index >= current.length) {
        return undefined;
      }
      current = current[index];
      continue;
    }

    if (!isRecord(current) || !(segment in current)) {
      return undefined;
    }

    current = current[segment];
  }

  return current;
}

function stringifyWasmJson(value: unknown): string {
  return JSON.stringify(value, (_key, candidate) => {
    if (typeof candidate === 'bigint') {
      return candidate.toString();
    }

    if (candidate instanceof Uint8Array) {
      return Array.from(candidate);
    }

    return candidate;
  });
}

async function hashInlineBytes(bytes: Uint8Array): Promise<string> {
  // Lazy import keeps the pre-compiled-Module path free of node:crypto.
  const { createHash } = await import('node:crypto');
  return createHash('sha256').update(bytes).digest('hex');
}

function decodeBase64Bytes(base64: string): Uint8Array {
  return Uint8Array.from(Buffer.from(base64, 'base64'));
}

function toWasmBinary(bytes: Uint8Array): ArrayBuffer {
  return bytes.buffer.slice(
    bytes.byteOffset,
    bytes.byteOffset + bytes.byteLength
  ) as ArrayBuffer;
}

function ensureWasmState(payload: unknown, nodeId: string): WasmBridgeState {
  if (
    payload &&
    typeof payload === 'object' &&
    'descriptor' in payload &&
    'locals' in payload
  ) {
    return payload as WasmBridgeState;
  }

  throw new Error(
    `WasmBridge node '${nodeId}' received invalid state. ` +
      `Expected { input, locals, descriptor }, got: ${typeof payload}`
  );
}

function coerceNonNegativeInteger(value: unknown, label: string): number {
  if (typeof value === 'number' && Number.isInteger(value) && value >= 0) {
    return value;
  }

  if (typeof value === 'bigint' && value >= 0n) {
    const coerced = Number(value);
    if (Number.isSafeInteger(coerced)) {
      return coerced;
    }
  }

  throw new Error(`${label} must be a non-negative integer.`);
}

function normalizeScalarArg(value: unknown): number | bigint {
  if (typeof value === 'number') {
    if (!Number.isFinite(value)) {
      throw new Error('scalar-v1 arguments must be finite numbers.');
    }
    return value;
  }

  if (typeof value === 'bigint') {
    return value;
  }

  if (typeof value === 'boolean') {
    return value ? 1 : 0;
  }

  if (typeof value === 'string') {
    const parsed = Number(value);
    if (!Number.isFinite(parsed)) {
      throw new Error(
        `scalar-v1 string arguments must parse to finite numbers. Received '${value}'.`
      );
    }
    return parsed;
  }

  throw new Error(
    `scalar-v1 arguments must be numbers, booleans, bigints, or numeric strings. Received ${typeof value}.`
  );
}

function normalizeScalarArgs(input: unknown): readonly (number | bigint)[] {
  if (input === undefined || input === null) {
    return [];
  }

  if (Array.isArray(input)) {
    return input.map((value) => normalizeScalarArg(value));
  }

  if (isRecord(input) && Array.isArray(input.args)) {
    return input.args.map((value) => normalizeScalarArg(value));
  }

  return [normalizeScalarArg(input)];
}

function pickString(
  ...values: readonly (string | undefined)[]
): string | undefined {
  for (const value of values) {
    if (typeof value === 'string' && value.length > 0) {
      return value;
    }
  }

  return undefined;
}

function normalizeAbi(
  value: string | WasmExecutionAbi | undefined
): WasmExecutionAbi {
  return value === 'scalar-v1' || value === 'json-v1'
    ? value
    : DEFAULT_WASM_ABI;
}

function resolveDescriptor(
  manifest: WasmExecutionManifest,
  stateDescriptor: Readonly<ResolvedWasmExecutionDescriptor> | undefined,
  plan: WasmNodeExecutionPlan | undefined,
  props: Record<string, string>
): ResolvedWasmExecutionDescriptor {
  const filePath = pickString(
    props.file,
    plan?.file_path,
    stateDescriptor?.filePath,
    manifest.file_path
  );
  const wasmBase64 = pickString(
    props.wasmBase64,
    props.wasm_base64,
    plan?.wasm_base64,
    stateDescriptor?.wasmBase64,
    manifest.wasm_base64
  );
  const wasmModule =
    plan?.wasm_module ?? stateDescriptor?.wasmModule ?? manifest.wasm_module;
  const artifactId = pickString(
    stateDescriptor?.artifactId,
    manifest.artifact_id
  );
  const exportName =
    pickString(
      props.export,
      props.exportName,
      props.callee,
      plan?.export_name,
      stateDescriptor?.exportName,
      manifest.export_name
    ) ?? DEFAULT_WASM_EXPORT_NAME;
  const abi = normalizeAbi(
    pickString(props.abi, plan?.abi, stateDescriptor?.abi, manifest.abi)
  );
  const memoryExport =
    pickString(
      props.memory,
      props.memoryExport,
      plan?.memory_export,
      stateDescriptor?.memoryExport,
      manifest.memory_export
    ) ?? DEFAULT_MEMORY_EXPORT;
  const allocExport =
    pickString(
      props.alloc,
      props.allocExport,
      plan?.alloc_export,
      stateDescriptor?.allocExport,
      manifest.alloc_export
    ) ?? DEFAULT_ALLOC_EXPORT;
  const resultLengthExport =
    pickString(
      props.resultLen,
      props.resultLength,
      props.resultLengthExport,
      plan?.result_length_export,
      stateDescriptor?.resultLengthExport,
      manifest.result_length_export
    ) ?? DEFAULT_RESULT_LENGTH_EXPORT;
  const freeExport =
    pickString(
      props.free,
      props.freeExport,
      plan?.free_export,
      stateDescriptor?.freeExport,
      manifest.free_export
    ) ?? DEFAULT_FREE_EXPORT;
  const resolveRelativeTo =
    stateDescriptor?.resolveRelativeTo ?? manifest.resolve_relative_to;

  return {
    ...(artifactId ? { artifactId } : {}),
    ...(filePath ? { filePath } : {}),
    ...(wasmBase64 ? { wasmBase64 } : {}),
    ...(wasmModule ? { wasmModule } : {}),
    exportName,
    abi,
    memoryExport,
    allocExport,
    resultLengthExport,
    ...(freeExport ? { freeExport } : {}),
    ...(resolveRelativeTo ? { resolveRelativeTo } : {}),
  };
}

function resolveDescriptorInput(
  state: WasmBridgeState,
  props: Record<string, string>,
  defaultValue: unknown
): unknown {
  const from = props.from?.trim();
  if (!from) {
    return defaultValue;
  }

  if (from === 'input') {
    return state.input;
  }

  if (from === 'locals') {
    return state.locals;
  }

  return resolveContextPath(
    {
      input: state.input,
      locals: state.locals,
    },
    from
  );
}

async function resolveFilesystemPath(
  filePath: string,
  relativeTo: string | undefined
): Promise<string> {
  // Lazy import keeps the pre-compiled-Module path free of node:path.
  const { dirname, isAbsolute, resolve } = await import('node:path');
  if (isAbsolute(filePath)) {
    return resolve(filePath);
  }

  if (!relativeTo) {
    return resolve(filePath);
  }

  return resolve(dirname(relativeTo), filePath);
}

function loadCachedModule(
  cacheKey: string,
  load: () => Promise<LoadedWasmModule>
): Promise<LoadedWasmModule> {
  const cached = wasmModuleCache.get(cacheKey);
  if (cached) {
    return cached;
  }

  const pending = load().catch((error) => {
    wasmModuleCache.delete(cacheKey);
    throw error;
  });
  wasmModuleCache.set(cacheKey, pending);
  return pending;
}

async function compileWasmModule(bytes: Uint8Array): Promise<LoadedWasmModule> {
  const module = new WebAssembly.Module(new Uint8Array(toWasmBinary(bytes)));
  return {
    module,
    imports: WebAssembly.Module.imports(module),
  };
}

function toLoadedWasmModule(
  value: LoadedWasmModule | WebAssembly.Module
): LoadedWasmModule {
  if (value instanceof WebAssembly.Module) {
    return {
      module: value,
      imports: WebAssembly.Module.imports(value),
    };
  }

  return value;
}

async function loadWasmModule(
  descriptor: Readonly<ResolvedWasmExecutionDescriptor>,
  hostOptions: WasmExecutionHostOptions
): Promise<LoadedWasmModule> {
  // Load order: moduleResolver → descriptor wasmModule → inline base64 →
  // file path. Pre-compiled Modules never go through
  // `new WebAssembly.Module(bytes)` (Cloudflare Workers forbid runtime
  // compile) and bypass the keyed byte cache — they are already compiled, so
  // there is no compile work to amortize.
  if (hostOptions.moduleResolver) {
    const resolved = await hostOptions.moduleResolver(descriptor);
    if (resolved) {
      return toLoadedWasmModule(resolved);
    }
  }

  if (descriptor.wasmModule) {
    return toLoadedWasmModule(descriptor.wasmModule);
  }

  if (descriptor.wasmBase64) {
    const bytes = decodeBase64Bytes(descriptor.wasmBase64);
    const cacheKey = `inline:${await hashInlineBytes(bytes)}`;
    return loadCachedModule(cacheKey, () => compileWasmModule(bytes));
  }

  if (!descriptor.filePath) {
    throw new Error(
      'WASM execution requires a module resolver match, a pre-compiled module, a file path, or inline base64 bytes.'
    );
  }

  // Lazy import keeps the pre-compiled-Module path free of node:fs.
  const { readFile, stat } = await import('node:fs/promises');
  const resolvedPath = await resolveFilesystemPath(
    descriptor.filePath,
    descriptor.resolveRelativeTo
  );
  const fileStats = await stat(resolvedPath);
  const cacheKey = `file:${resolvedPath}:${fileStats.size}:${fileStats.mtimeMs}`;
  return loadCachedModule(cacheKey, async () => {
    const bytes = new Uint8Array(await readFile(resolvedPath));
    return compileWasmModule(bytes);
  });
}

async function resolveImportObject(
  descriptor: Readonly<ResolvedWasmExecutionDescriptor>,
  hostOptions: WasmExecutionHostOptions,
  moduleImports: readonly WebAssembly.ModuleImportDescriptor[]
): Promise<WebAssembly.Imports> {
  if (moduleImports.length === 0) {
    return {};
  }

  if (!hostOptions.importObject) {
    const importList = moduleImports
      .map((entry) => `${entry.module}.${entry.name}`)
      .join(', ');

    if (
      moduleImports.some((entry) => entry.module === 'wasi_snapshot_preview1')
    ) {
      throw new Error(
        `WASI guest '${descriptor.exportName}' requires a dedicated WASI host. Missing imports: ${importList}`
      );
    }

    throw new Error(
      `WASM guest '${descriptor.exportName}' requires host imports that were not provided: ${importList}`
    );
  }

  if (typeof hostOptions.importObject === 'function') {
    return await hostOptions.importObject(descriptor);
  }

  return hostOptions.importObject;
}

function getExportedFunction(
  exportsObject: WebAssembly.Exports,
  exportName: string
): (...args: readonly unknown[]) => unknown {
  const exported = exportsObject[exportName];
  if (typeof exported !== 'function') {
    throw new Error(`WASM export '${exportName}' is not a callable function.`);
  }

  return exported as (...args: readonly unknown[]) => unknown;
}

function getExportedMemory(
  exportsObject: WebAssembly.Exports,
  exportName: string
): WebAssembly.Memory {
  const exported = exportsObject[exportName];
  if (!(exported instanceof WebAssembly.Memory)) {
    throw new Error(`WASM export '${exportName}' is not a memory export.`);
  }

  return exported;
}

async function executeScalarDescriptor(
  descriptor: Readonly<ResolvedWasmExecutionDescriptor>,
  input: unknown,
  hostOptions: WasmExecutionHostOptions
): Promise<unknown> {
  const loaded = await loadWasmModule(descriptor, hostOptions);
  const imports = await resolveImportObject(
    descriptor,
    hostOptions,
    loaded.imports
  );
  const instance = await WebAssembly.instantiate(loaded.module, imports);
  const exportedFn = getExportedFunction(
    instance.exports,
    descriptor.exportName
  );
  return exportedFn(...normalizeScalarArgs(input));
}

async function executeJsonDescriptor(
  descriptor: Readonly<ResolvedWasmExecutionDescriptor>,
  input: unknown,
  hostOptions: WasmExecutionHostOptions
): Promise<unknown> {
  const loaded = await loadWasmModule(descriptor, hostOptions);
  const imports = await resolveImportObject(
    descriptor,
    hostOptions,
    loaded.imports
  );
  const instance = await WebAssembly.instantiate(loaded.module, imports);
  const exportsObject = instance.exports;
  const exportedFn = getExportedFunction(exportsObject, descriptor.exportName);
  const alloc = getExportedFunction(exportsObject, descriptor.allocExport);
  const getResultLength = getExportedFunction(
    exportsObject,
    descriptor.resultLengthExport
  );
  const memory = getExportedMemory(exportsObject, descriptor.memoryExport);
  const free =
    descriptor.freeExport &&
    typeof exportsObject[descriptor.freeExport] === 'function'
      ? (exportsObject[descriptor.freeExport] as (
          ...args: readonly unknown[]
        ) => unknown)
      : null;

  const inputBytes = new TextEncoder().encode(stringifyWasmJson(input));
  const inputPtr = coerceNonNegativeInteger(
    alloc(inputBytes.length),
    `${descriptor.allocExport} return value`
  );
  new Uint8Array(memory.buffer, inputPtr, inputBytes.length).set(inputBytes);

  const outputPtr = coerceNonNegativeInteger(
    exportedFn(inputPtr, inputBytes.length),
    `${descriptor.exportName} return value`
  );
  const outputLen = coerceNonNegativeInteger(
    getResultLength(),
    `${descriptor.resultLengthExport} return value`
  );
  const outputBytes = new Uint8Array(
    memory.buffer,
    outputPtr,
    outputLen
  ).slice();

  if (free) {
    free(inputPtr, inputBytes.length);
    free(outputPtr, outputLen);
  }

  const outputText = new TextDecoder('utf-8', { fatal: true }).decode(
    outputBytes
  );
  if (outputText.trim().length === 0) {
    return null;
  }

  try {
    return JSON.parse(outputText) as unknown;
  } catch (error) {
    throw new Error(
      `WASM guest '${descriptor.exportName}' returned invalid JSON: ${
        error instanceof Error ? error.message : String(error)
      }\n${outputText}`
    );
  }
}

async function executeWasmDescriptor(
  descriptor: Readonly<ResolvedWasmExecutionDescriptor>,
  input: unknown,
  hostOptions: WasmExecutionHostOptions
): Promise<unknown> {
  return descriptor.abi === 'scalar-v1'
    ? executeScalarDescriptor(descriptor, input, hostOptions)
    : executeJsonDescriptor(descriptor, input, hostOptions);
}

/**
 * Creates the Wasm Bridge Execution Ast.
 */
export function createWasmBridgeExecutionAst(
  assignTo = DEFAULT_ASSIGN_TO
): GraphAST {
  const nodes = new Map<string, ASTNode>([
    [
      'wasm_entry',
      {
        id: 'wasm_entry',
        labels: ['Entry', 'WasmBridgeEntry'],
        properties: {},
      },
    ],
    [
      'wasm_call',
      {
        id: 'wasm_call',
        labels: ['Statement', 'WasmBridgeCall'],
        properties: {
          assignTo,
        },
      },
    ],
    [
      'wasm_return',
      {
        id: 'wasm_return',
        labels: ['Return', 'WasmBridgeReturn'],
        properties: {
          returnKey: assignTo,
        },
      },
    ],
  ]);
  const edges: ASTEdge[] = [
    {
      sourceIds: ['wasm_entry'],
      targetIds: ['wasm_call'],
      type: 'PROCESS',
      properties: {},
    },
    {
      sourceIds: ['wasm_call'],
      targetIds: ['wasm_return'],
      type: 'PROCESS',
      properties: {},
    },
  ];

  return {
    nodes,
    edges,
  };
}

/**
 * Handles the gnosis register Wasm Bridge Handlers workflow.
 */
export function registerWasmBridgeHandlers(
  registry: GnosisRegistry,
  manifest: WasmExecutionManifest,
  hostOptions: WasmExecutionHostOptions = {}
): void {
  const plansByNodeId = new Map(
    manifest.node_execution_plans.map((plan) => [plan.node_id, plan] as const)
  );

  registry.register('WasmBridgeEntry', async (payload, props, context) => {
    const descriptor = resolveDescriptor(
      manifest,
      undefined,
      context?.nodeId ? plansByNodeId.get(context.nodeId) : undefined,
      props
    );
    return {
      input: payload,
      locals: {},
      descriptor,
    } satisfies WasmBridgeState;
  });

  registry.register('WasmBridgeCall', async (payload, props, context) => {
    const nodeId = context?.nodeId;
    if (!nodeId) {
      throw new Error('WasmBridgeCall requires a node id.');
    }

    const state = ensureWasmState(payload, nodeId);
    const plan = plansByNodeId.get(nodeId);
    const descriptor = resolveDescriptor(
      manifest,
      state.descriptor,
      plan,
      props
    );
    const callInput = resolveDescriptorInput(state, props, state.input);
    const value = await executeWasmDescriptor(
      descriptor,
      callInput,
      hostOptions
    );
    const assignTo = props.assignTo?.trim() || descriptor.exportName;

    return {
      input: state.input,
      locals: {
        ...state.locals,
        [assignTo]: value,
      },
      descriptor,
    } satisfies WasmBridgeState;
  });

  registry.register('WasmBridgeStatement', async (payload, props, context) => {
    const nodeId = context?.nodeId;
    if (!nodeId) {
      throw new Error('WasmBridgeStatement requires a node id.');
    }

    const state = ensureWasmState(payload, nodeId);
    const plan = plansByNodeId.get(nodeId);
    const descriptor = resolveDescriptor(
      manifest,
      state.descriptor,
      plan,
      props
    );
    const statementInput = resolveDescriptorInput(state, props, state.locals);
    const value = await executeWasmDescriptor(
      descriptor,
      statementInput,
      hostOptions
    );

    return {
      input: state.input,
      locals: {
        ...state.locals,
        __last__: value,
      },
      descriptor,
    } satisfies WasmBridgeState;
  });

  registry.register('WasmBridgeReturn', async (payload, props, context) => {
    const nodeId = context?.nodeId;
    if (!nodeId) {
      throw new Error('WasmBridgeReturn requires a node id.');
    }

    const state = ensureWasmState(payload, nodeId);
    const returnKey = props.returnKey?.trim();
    if (returnKey && returnKey in state.locals) {
      return state.locals[returnKey];
    }

    if ('__last__' in state.locals) {
      return state.locals.__last__;
    }

    return state.locals;
  });

  registry.register('WasmBridgeJoin', async (payload, _props, context) => {
    const nodeId = context?.nodeId;
    if (!nodeId) {
      throw new Error('WasmBridgeJoin requires a node id.');
    }

    if (!Array.isArray(payload)) {
      return payload;
    }

    const merged: WasmBridgeState = {
      input: payload[0]?.input,
      locals: {},
      descriptor:
        payload[0]?.descriptor ??
        resolveDescriptor(manifest, undefined, undefined, {}),
    };

    for (const branch of payload) {
      if (branch && typeof branch === 'object' && 'locals' in branch) {
        Object.assign(merged.locals, (branch as WasmBridgeState).locals);
      }
    }

    return merged;
  });
}

/**
 * Handles the gnosis execute Wasm With Gnosis workflow.
 */
export async function executeWasmWithGnosis(
  options: ExecuteWasmOptions
): Promise<GnosisEngineExecutionResult> {
  const registry = options.registry ?? new GnosisRegistry();
  registerWasmBridgeHandlers(registry, options.manifest, options);
  const engine = new GnosisEngine(registry, options.engineOptions ?? {});
  return engine.executeWithResult(
    options.ast,
    options.input,
    options.executeOptions ?? {}
  );
}
