// @a0n/gnosis/src/weather-mesh-adapter.ts
/**
 * Weather Mesh Adapter
 *
 * Engine-agnostic bridge that turns a storms-ingest mesh selection into
 * scheduling-ready task units with O(1) neighbor traversal.
 *
 * The adapter is structurally typed against aeon-weather's `MeshCoordinate`
 * shape so Gnosis does not take a direct dependency on the weather package.
 * Any module that produces objects with the same shape can feed this
 * adapter -- the contract is "cells carry canonical FRF-64 hex IDs and an
 * (li, oi) index pair".
 *
 * # What it does
 *
 * Given a list of mesh cells (each with a spatial hash, centroid, and active
 * flag), the adapter:
 *
 * 1. Builds a primary `Map<string, WeatherMeshTaskUnit>` keyed by
 *    `spatialHex64`. This is the O(1) lookup surface.
 * 2. Links each task unit to its 8 spatial neighbors and its t±1 temporal
 *    neighbors by *ID*, not by index. Neighbor stepping is a single Map
 *    lookup per hop; no lookup-table-wide scans.
 * 3. Assigns a scheduling priority via god-formula weight: active cells
 *    inherit full weight, dormant cells are demoted.
 *
 * # Why
 *
 * The scheduling mesh inside `scheduling-mesh.ts` is a topology-agnostic
 * k-gon tiling. This adapter is the thin translation layer that lets weather
 * cells plug into that topology without inventing a parallel graph: the
 * scheduler sees uniform task units, and fork/race/fold work unchanged.
 */

/**
 * Minimal structural type for a mesh coordinate as emitted by aeon-weather.
 * We only take the fields the adapter needs; callers can pass the full
 * `MeshCoordinate` object unchanged.
 */
export interface WeatherMeshCellLike {
  readonly latIndex: number;
  readonly lonIndex: number;
  readonly spatialHex64: string;
  readonly spaceTimeHex64?: string;
  readonly timeIndex?: number;
  readonly centroid?: { readonly lat: number; readonly lon: number };
  readonly areaWeightBps?: number;
  readonly active?: boolean;
  readonly activeReason?: string;
}

/** Neighbor reference table: eight spatial + three temporal IDs. */
export interface WeatherMeshNeighborRefs {
  readonly nw: string | null;
  readonly n: string | null;
  readonly ne: string | null;
  readonly w: string | null;
  readonly e: string | null;
  readonly sw: string | null;
  readonly s: string | null;
  readonly se: string | null;
  readonly tPrev: string | null;
  readonly tNext: string | null;
}

/** A single scheduling task unit derived from a mesh cell. */
export interface WeatherMeshTaskUnit {
  /** Primary ID: the canonical FRF-64 spatial hash. */
  readonly id: string;
  /** Integer grid indices for locality-preserving neighbor computation. */
  readonly latIndex: number;
  readonly lonIndex: number;
  /** Optional hour-bucket index when the task spans a time axis. */
  readonly timeIndex: number | null;
  /** Snapped grid centroid. */
  readonly centroid: { readonly lat: number; readonly lon: number } | null;
  /** Scheduling weight. Active cells = `areaWeightBps`, dormant = 0. */
  readonly weight: number;
  /** Whether the cell is in the active mask. */
  readonly active: boolean;
  /** Human-readable reason the cell is active (or 'dormant'). */
  readonly activeReason: string;
  /** Pre-resolved neighbor IDs for O(1) stepping. */
  readonly neighbors: WeatherMeshNeighborRefs;
}

/** Internal: 721 / 1440 at 0.25° -- kept here so the adapter is self-contained. */
const DEFAULT_LAT_INDEX_COUNT = 721;
const DEFAULT_LON_INDEX_COUNT = 1440;

/**
 * Transform a flat list of mesh cells into a dense `Map` of scheduling task
 * units with bidirectional neighbor references.
 *
 * Because the caller provides cells that already carry canonical hex IDs,
 * the adapter can resolve most neighbors just by looking up the hex from the
 * input index. Neighbors that fall outside the supplied set (e.g., edge of a
 * bbox) are still listed by ID using the same hash function, so the
 * scheduler can *discover* them at run time without a lookup-table scan.
 *
 * @param cells - Mesh cells to schedule. Order is irrelevant.
 * @param options.hashSpatial - Pure function that returns the canonical
 *   spatial hash for any (lat, lon) grid index pair. Defaults to reading the
 *   hex off the input cell when possible; required when neighbors fall
 *   outside the input set.
 * @param options.latIndexCount / lonIndexCount - Grid dimensions. Defaults to
 *   the 0.25° globe (721 × 1440).
 */
export function buildWeatherMeshTaskUnits(
  cells: readonly WeatherMeshCellLike[],
  options: {
    hashSpatial: (latIdx: number, lonIdx: number) => string;
    latIndexCount?: number;
    lonIndexCount?: number;
    hashSpaceTime?: (
      latIdx: number,
      lonIdx: number,
      timeIdx: number
    ) => string;
  }
): Map<string, WeatherMeshTaskUnit> {
  const latMax = (options.latIndexCount ?? DEFAULT_LAT_INDEX_COUNT) - 1;
  const lonMod = options.lonIndexCount ?? DEFAULT_LON_INDEX_COUNT;
  const byCoord = new Map<string, WeatherMeshCellLike>();
  const output = new Map<string, WeatherMeshTaskUnit>();

  for (const cell of cells) {
    byCoord.set(coordKey(cell.latIndex, cell.lonIndex), cell);
  }

  for (const cell of cells) {
    const li = cell.latIndex;
    const oi = cell.lonIndex;

    const north = Math.min(latMax, li + 1);
    const south = Math.max(0, li - 1);
    const east = (oi + 1) % lonMod;
    const west = (oi - 1 + lonMod) % lonMod;

    const neighbors: WeatherMeshNeighborRefs = {
      nw: neighborHash(byCoord, options.hashSpatial, north, west),
      n: neighborHash(byCoord, options.hashSpatial, north, oi),
      ne: neighborHash(byCoord, options.hashSpatial, north, east),
      w: neighborHash(byCoord, options.hashSpatial, li, west),
      e: neighborHash(byCoord, options.hashSpatial, li, east),
      sw: neighborHash(byCoord, options.hashSpatial, south, west),
      s: neighborHash(byCoord, options.hashSpatial, south, oi),
      se: neighborHash(byCoord, options.hashSpatial, south, east),
      tPrev:
        options.hashSpaceTime && cell.timeIndex !== undefined
          ? options.hashSpaceTime(li, oi, cell.timeIndex - 1)
          : null,
      tNext:
        options.hashSpaceTime && cell.timeIndex !== undefined
          ? options.hashSpaceTime(li, oi, cell.timeIndex + 1)
          : null,
    };

    const active = cell.active !== false;
    const weight = active ? cell.areaWeightBps ?? 0 : 0;

    const unit: WeatherMeshTaskUnit = {
      id: cell.spatialHex64,
      latIndex: li,
      lonIndex: oi,
      timeIndex: cell.timeIndex ?? null,
      centroid: cell.centroid ?? null,
      weight,
      active,
      activeReason: cell.activeReason ?? (active ? 'active' : 'dormant'),
      neighbors,
    };
    output.set(unit.id, unit);
  }

  return output;
}

/**
 * O(1) neighbor lookup: given a task unit and a direction, return the
 * neighbor unit from the map, or `null` when the neighbor is outside the
 * current selection. The ID is always resolved (from the pre-computed
 * `neighbors` table), so this never scans the mesh.
 */
export function stepWeatherMeshNeighbor(
  units: Map<string, WeatherMeshTaskUnit>,
  unit: WeatherMeshTaskUnit,
  direction: keyof WeatherMeshNeighborRefs
): WeatherMeshTaskUnit | null {
  const id = unit.neighbors[direction];
  if (!id) {
    return null;
  }
  return units.get(id) ?? null;
}

/**
 * Iterate the active-only task units in deterministic order (sorted by
 * `latIndex` then `lonIndex`). Useful for fork/race fan-out.
 */
export function iterateActiveWeatherMeshUnits(
  units: Map<string, WeatherMeshTaskUnit>
): WeatherMeshTaskUnit[] {
  const active: WeatherMeshTaskUnit[] = [];
  for (const unit of units.values()) {
    if (unit.active) {
      active.push(unit);
    }
  }
  active.sort((left, right) =>
    left.latIndex === right.latIndex
      ? left.lonIndex - right.lonIndex
      : left.latIndex - right.latIndex
  );
  return active;
}

function coordKey(latIdx: number, lonIdx: number): string {
  return `${latIdx}:${lonIdx}`;
}

function neighborHash(
  byCoord: Map<string, WeatherMeshCellLike>,
  hashSpatial: (latIdx: number, lonIdx: number) => string,
  latIdx: number,
  lonIdx: number
): string {
  const known = byCoord.get(coordKey(latIdx, lonIdx));
  if (known) {
    return known.spatialHex64;
  }
  return hashSpatial(latIdx, lonIdx);
}
