// @a0n/gnosis/src/worldline.ts
/**
 * Worldline Traces — trajectories through spacetime.
 *
 * A worldline records the path a body takes through spacetime,
 * accumulating proper time along the way. Stationary bodies,
 * constant-velocity travelers, and orbiting bodies each produce
 * distinct worldline shapes with different proper-time accumulation
 * due to time dilation.
 *
 * All spatial coordinates are in light-years; all times are in
 * picolorenzos. Proper time is computed from the spacetime interval
 * between consecutive points.
 *
 * @packageDocumentation
 */

import type { SpacetimeEvent, PenroseCoordinate } from './causal-topology';
import { spacetimeInterval, penroseCoordinates } from './causal-topology';
import { LIGHT_PICOLORENZO_LY } from './math';

// ─── Types ─────────────────────────────────────────────────────────

export interface WorldlinePoint {
  readonly pLo: number;
  readonly x: number;  // light-years
  readonly y: number;
  readonly z: number;
  readonly properTimePLo: number;  // accumulated proper time along worldline
}

export interface WorldlineTrace {
  readonly source: 'GnosisWorldline';
  readonly bodyId: string;
  readonly label: string;
  readonly points: readonly WorldlinePoint[];
  readonly totalProperTimePLo: number;
  readonly totalCoordinateTimePLo: number;
  readonly startPLo: number;
  readonly endPLo: number;
}

export interface WorldlinePenroseTrace {
  readonly bodyId: string;
  readonly points: readonly PenroseCoordinate[];
}

// ─── Core Functions ────────────────────────────────────────────────

/**
 * Build a worldline trace from time-ordered positions.
 *
 * Computes proper time between consecutive points using spacetimeInterval
 * and accumulates it along the path. The first point has properTimePLo = 0.
 */
export function worldlineFromPositions(
  bodyId: string,
  label: string,
  positions: readonly { pLo: number; x: number; y: number; z: number }[]
): WorldlineTrace {
  if (positions.length === 0) {
    return {
      source: 'GnosisWorldline',
      bodyId,
      label,
      points: [],
      totalProperTimePLo: 0,
      totalCoordinateTimePLo: 0,
      startPLo: 0,
      endPLo: 0,
    };
  }

  const points: WorldlinePoint[] = [];
  let accumulatedProperTime = 0;

  points.push({
    pLo: positions[0].pLo,
    x: positions[0].x,
    y: positions[0].y,
    z: positions[0].z,
    properTimePLo: 0,
  });

  for (let i = 1; i < positions.length; i++) {
    const prev = positions[i - 1];
    const curr = positions[i];

    const fromEvent: SpacetimeEvent = {
      id: `${bodyId}-${i - 1}`,
      label: `${label} @ ${i - 1}`,
      x: prev.x,
      y: prev.y,
      z: prev.z,
      pLo: prev.pLo,
    };

    const toEvent: SpacetimeEvent = {
      id: `${bodyId}-${i}`,
      label: `${label} @ ${i}`,
      x: curr.x,
      y: curr.y,
      z: curr.z,
      pLo: curr.pLo,
    };

    const interval = spacetimeInterval(fromEvent, toEvent);

    // For timelike intervals, proper time = sqrt(-ds^2)
    // For lightlike/spacelike, proper time contribution is 0
    if (interval.properTimePLo !== null) {
      accumulatedProperTime += interval.properTimePLo;
    }

    points.push({
      pLo: curr.pLo,
      x: curr.x,
      y: curr.y,
      z: curr.z,
      properTimePLo: accumulatedProperTime,
    });
  }

  const startPLo = positions[0].pLo;
  const endPLo = positions[positions.length - 1].pLo;

  return {
    source: 'GnosisWorldline',
    bodyId,
    label,
    points,
    totalProperTimePLo: accumulatedProperTime,
    totalCoordinateTimePLo: endPLo - startPLo,
    startPLo,
    endPLo,
  };
}

/**
 * Convert a worldline trace to Penrose diagram coordinates.
 *
 * Each worldline point becomes a SpacetimeEvent, then penroseCoordinates
 * maps it into the conformal diamond.
 */
export function worldlineToPenrose(
  trace: WorldlineTrace,
  origin: SpacetimeEvent
): WorldlinePenroseTrace {
  const events: SpacetimeEvent[] = trace.points.map((pt, i) => ({
    id: `${trace.bodyId}-${i}`,
    label: `${trace.label} @ ${i}`,
    x: pt.x,
    y: pt.y,
    z: pt.z,
    pLo: pt.pLo,
  }));

  const penrose = penroseCoordinates(events, origin);

  return {
    bodyId: trace.bodyId,
    points: penrose,
  };
}

// ─── Generators ────────────────────────────────────────────────────

/**
 * A body sitting still. All points have the same x, y, z but advancing pLo.
 * Proper time equals coordinate time for stationary bodies.
 */
export function stationaryWorldline(
  bodyId: string,
  label: string,
  position: { x: number; y: number; z: number },
  startPLo: number,
  endPLo: number,
  steps: number = 100
): WorldlineTrace {
  const positions: { pLo: number; x: number; y: number; z: number }[] = [];
  const dt = (endPLo - startPLo) / steps;

  for (let i = 0; i <= steps; i++) {
    positions.push({
      pLo: startPLo + i * dt,
      x: position.x,
      y: position.y,
      z: position.z,
    });
  }

  return worldlineFromPositions(bodyId, label, positions);
}

/**
 * Constant velocity worldline. Position advances linearly.
 * Proper time accumulates with time dilation: tau < t for v > 0.
 *
 * Velocity components are in light-years per picolorenzo.
 */
export function uniformMotionWorldline(
  bodyId: string,
  label: string,
  startPos: { x: number; y: number; z: number },
  velocityLyPerPLo: { vx: number; vy: number; vz: number },
  startPLo: number,
  endPLo: number,
  steps: number = 100
): WorldlineTrace {
  const positions: { pLo: number; x: number; y: number; z: number }[] = [];
  const dt = (endPLo - startPLo) / steps;

  for (let i = 0; i <= steps; i++) {
    const t = i * dt;
    positions.push({
      pLo: startPLo + t,
      x: startPos.x + velocityLyPerPLo.vx * t,
      y: startPos.y + velocityLyPerPLo.vy * t,
      z: startPos.z + velocityLyPerPLo.vz * t,
    });
  }

  return worldlineFromPositions(bodyId, label, positions);
}

/**
 * Circular orbit in the x-z plane (matching aeon-solarium's ecliptic convention).
 *
 * x = centerX + radiusLy * cos(angle)
 * z = centerZ + radiusLy * sin(angle)
 * y = 0
 *
 * Angle advances as 2*pi*(pLo - startPLo) / periodPLo.
 */
export function circularOrbitWorldline(
  bodyId: string,
  label: string,
  centerX: number,
  centerZ: number,
  radiusLy: number,
  periodPLo: number,
  startPLo: number,
  endPLo: number,
  steps: number = 360
): WorldlineTrace {
  const positions: { pLo: number; x: number; y: number; z: number }[] = [];
  const dt = (endPLo - startPLo) / steps;

  for (let i = 0; i <= steps; i++) {
    const pLo = startPLo + i * dt;
    const angle = (2 * Math.PI * (pLo - startPLo)) / periodPLo;
    positions.push({
      pLo,
      x: centerX + radiusLy * Math.cos(angle),
      y: 0,
      z: centerZ + radiusLy * Math.sin(angle),
    });
  }

  return worldlineFromPositions(bodyId, label, positions);
}
