// @a0n/gnosis/forkjoin-app/client/components/Presence/PresenceFlightDeck.tsx
import { memo, useMemo, useState } from '@a0n/raect';
import {
  CollaborativePresenceScrollContainer,
  PresenceElementsPanel,
  type PresenceScroll,
  type PresenceScrollThemeTokens,
  type PresenceUser,
} from '../../shims/aeon-flux-react-presence-kit';
import type { Visitor } from '../../hooks/usePresence';

interface PresenceFlightDeckProps {
  visitors: Visitor[];
  localVisitorId: string;
  isConnected: boolean;
  onScrollStateChange?: (scroll: PresenceScroll) => void;
}

const CURSOR_LAYER_WIDTH = 320;
const CURSOR_LAYER_HEIGHT = 170;
const PRESENCE_SCROLL_ACCENT = 'var(--color-accent)';
const PRESENCE_SCROLL_MARKER_LIMIT = 40;
const PRESENCE_SCROLL_THEME: PresenceScrollThemeTokens = {
  accentColor: PRESENCE_SCROLL_ACCENT,
  railSurfaceColor: 'rgba(13, 17, 23, 0.92)',
  railBorderColor: 'rgba(70, 126, 170, 0.28)',
  textColor: '#e2edf8',
  mutedTextColor: '#9eb3c7',
};

export const PresenceFlightDeck = memo(function PresenceFlightDeck({
  visitors,
  localVisitorId,
  isConnected,
  onScrollStateChange,
}: PresenceFlightDeckProps) {
  const [isOpen, setIsOpen] = useState(false);

  const presenceUsers = useMemo<PresenceUser[]>(() => {
    const nowIso = new Date().toISOString();
    return visitors.map((visitor) => ({
      userId: visitor.id,
      role: 'user',
      status: 'online',
      lastActivity: visitor.lastActivityAt
        ? new Date(visitor.lastActivityAt).toISOString()
        : nowIso,
      cursor: visitor.cursor
        ? {
            x: visitor.cursor.x * CURSOR_LAYER_WIDTH,
            y: visitor.cursor.y * CURSOR_LAYER_HEIGHT,
          }
        : undefined,
      focusNode: visitor.focusNode,
      selection: visitor.selection,
      typing: visitor.typing,
      scroll: visitor.scroll,
      viewport: visitor.viewport,
      inputState: visitor.inputState,
      emotion: visitor.emotion,
      editing: visitor.editing,
    }));
  }, [visitors]);

  return (
    <div className="presence-flight-deck">
      <button
        type="button"
        className="presence-flight-deck-toggle"
        onClick={() => setIsOpen((prev) => !prev)}
        aria-expanded={isOpen}
        aria-label="Toggle presence flight deck"
      >
        <span
          className={`presence-flight-deck-dot ${
            isConnected ? 'live' : 'offline'
          }`}
        />
        Presence
      </button>

      {isOpen ? (
        <aside
          className="presence-flight-deck-panel"
          role="complementary"
          aria-label="Presence flight deck"
        >
          <header className="presence-flight-deck-header">
            <strong>Presence Flight Deck</strong>
            <small>{presenceUsers.length} active</small>
          </header>
          <CollaborativePresenceScrollContainer
            presence={presenceUsers}
            localUserId={localVisitorId}
            height={430}
            onScrollStateChange={onScrollStateChange}
            className="presence-flight-deck-scroll"
            accentColor={PRESENCE_SCROLL_ACCENT}
            markerLimit={PRESENCE_SCROLL_MARKER_LIMIT}
            theme={PRESENCE_SCROLL_THEME}
            style={{
              background:
                'linear-gradient(180deg, rgba(10, 13, 18, 0.95), rgba(13, 17, 23, 0.92))',
              border: '1px solid rgba(70, 126, 170, 0.28)',
              boxShadow:
                'inset 0 0 0 1px rgba(114, 160, 201, 0.08), inset 0 14px 24px rgba(0, 0, 0, 0.24)',
            }}
          >
            <PresenceElementsPanel
              presence={presenceUsers}
              localUserId={localVisitorId}
              cursorLayerHeight={CURSOR_LAYER_HEIGHT}
              scrollBarAccentColor={PRESENCE_SCROLL_ACCENT}
              scrollBarMarkerLimit={PRESENCE_SCROLL_MARKER_LIMIT}
              scrollTheme={PRESENCE_SCROLL_THEME}
            />
          </CollaborativePresenceScrollContainer>
        </aside>
      ) : null}

      <style>{`
        .presence-flight-deck {
          position: fixed;
          right: 24px;
          bottom: 24px;
          z-index: 1200;
        }

        .presence-flight-deck-toggle {
          display: inline-flex;
          align-items: center;
          gap: 8px;
          border: 1px solid rgba(70, 126, 170, 0.4);
          background: rgba(8, 11, 16, 0.9);
          color: #f8fafc;
          border-radius: 999px;
          padding: 8px 14px;
          font-size: 0.82rem;
          font-weight: 600;
          cursor: pointer;
          backdrop-filter: blur(10px);
        }

        .presence-flight-deck-dot {
          width: 8px;
          height: 8px;
          border-radius: 50%;
          background: #6b8eaa;
        }

        .presence-flight-deck-dot.live {
          background: #10b981;
          box-shadow: 0 0 8px rgba(16, 185, 129, 0.8);
        }

        .presence-flight-deck-dot.offline {
          background: #ef4444;
        }

        .presence-flight-deck-panel {
          margin-top: 10px;
          width: min(380px, calc(100vw - 32px));
          border-radius: 14px;
          border: 1px solid rgba(70, 126, 170, 0.35);
          background: rgba(8, 11, 16, 0.9);
          backdrop-filter: blur(10px);
          overflow: hidden;
        }

        .presence-flight-deck-header {
          display: flex;
          align-items: center;
          justify-content: space-between;
          padding: 10px 12px;
          color: #f8fafc;
          border-bottom: 1px solid rgba(70, 126, 170, 0.24);
        }

        .presence-flight-deck-header small {
          color: rgba(226, 232, 240, 0.8);
          font-size: 0.76rem;
        }

        .presence-flight-deck-scroll {
          border: 0 !important;
          background: transparent !important;
        }

        .presence-flight-deck .aeon-collab-scroll-rail {
          backdrop-filter: blur(3px);
          transition: box-shadow 160ms ease;
        }

        .presence-flight-deck .aeon-collab-scroll-rail:focus-visible {
          outline: 2px solid rgba(114, 160, 201, 0.5);
          outline-offset: 2px;
          box-shadow:
            0 0 0 2px rgba(8, 11, 16, 0.75),
            0 0 12px rgba(114, 160, 201, 0.35);
        }

        .presence-flight-deck .aeon-collab-scroll-marker {
          transition: transform 140ms ease, filter 140ms ease;
        }

        .presence-flight-deck .aeon-collab-scroll-marker:hover {
          transform: translate(-50%, -50%) scale(1.15);
          filter: saturate(1.08) brightness(1.06);
        }
      `}</style>
    </div>
  );
});
