// @a0n/gnosis/forkjoin-app/client/components/Relay3D/MeshVisualization.tsx
/**
 * MeshVisualization - 3D visualization of Relay peer mesh networking
 *
 * Shows interconnected edge nodes with animated data flows.
 */

import { useRef, useMemo, useState } from '@a0n/raect';
import { Canvas, useFrame } from '@a0n/aeon-3d/fiber';
import { Text, Sphere, Float, Line } from '@a0n/aeon-3d/drei';
import * as THREE from '@a0n/aeon-3d/three';
import { SafeBloom } from '../3D/SafeEffects';

const CYAN = '#06b6d4';

interface EdgeNode {
  name: string;
  position: [number, number, number];
}

const edgeNodes: EdgeNode[] = [
  { name: 'Edge A', position: [-2, 1.5, 0] },
  { name: 'Edge B', position: [2, 1.5, 0] },
  { name: 'Edge C', position: [-2, -1.5, 0] },
  { name: 'Edge D', position: [2, -1.5, 0] },
];

// Connections between nodes (all-to-all mesh)
const connections: [number, number][] = [
  [0, 1], // A-B
  [0, 2], // A-C
  [0, 3], // A-D (diagonal)
  [1, 2], // B-C (diagonal)
  [1, 3], // B-D
  [2, 3], // C-D
];

/**
 * Renders the Mesh Visualization view.
 */
export function MeshVisualization() {
  return (
    <div style={{ width: '100%', height: '250px' }}>
      <Canvas camera={{ position: [0, 0, 7], fov: 45 }}>
        <ambientLight intensity={0.3} />
        <pointLight position={[10, 10, 10]} intensity={0.5} />
        <pointLight position={[-10, -10, 10]} intensity={0.3} color={CYAN} />

        <MeshScene />

        <SafeBloom
          intensity={0.5}
          luminanceThreshold={0.2}
          luminanceSmoothing={0.9}
        />
      </Canvas>
    </div>
  );
}

function MeshScene() {
  const groupRef = useRef<THREE.Group>(null);

  useFrame((state) => {
    if (!groupRef.current) return;
    const time = state.clock.elapsedTime;
    groupRef.current.rotation.y = Math.sin(time * 0.15) * 0.15;
    groupRef.current.rotation.x = Math.sin(time * 0.1) * 0.05;
  });

  return (
    <group ref={groupRef}>
      {/* Edge Nodes */}
      {edgeNodes.map((node, i) => (
        <EdgeNodeSphere key={node.name} node={node} index={i} />
      ))}

      {/* Mesh Connections */}
      {connections.map(([from, to], i) => (
        <MeshConnection
          key={`${from}-${to}`}
          from={edgeNodes[from].position}
          to={edgeNodes[to].position}
          index={i}
        />
      ))}

      {/* Data packets flowing */}
      {connections.map(([from, to], i) => (
        <DataPacket
          key={`packet-${from}-${to}`}
          from={edgeNodes[from].position}
          to={edgeNodes[to].position}
          delay={i * 0.5}
        />
      ))}
    </group>
  );
}

interface EdgeNodeSphereProps {
  node: EdgeNode;
  index: number;
}

function EdgeNodeSphere({ node, index }: EdgeNodeSphereProps) {
  const meshRef = useRef<THREE.Mesh>(null);

  useFrame((state) => {
    if (!meshRef.current) return;
    const time = state.clock.elapsedTime;

    // Pulse animation
    const scale = 0.3 + Math.sin(time * 2 + index) * 0.05;
    meshRef.current.scale.setScalar(scale);

    // Glow intensity
    const material = meshRef.current.material as THREE.MeshStandardMaterial;
    material.emissiveIntensity = 0.5 + Math.sin(time * 3 + index) * 0.2;
  });

  return (
    <Float speed={2} floatIntensity={0.1}>
      <group position={node.position}>
        <Sphere ref={meshRef} args={[1, 32, 32]}>
          <meshStandardMaterial
            color={CYAN}
            emissive={CYAN}
            emissiveIntensity={0.5}
            metalness={0.5}
            roughness={0.3}
          />
        </Sphere>
        <Text
          position={[0, -0.6, 0]}
          fontSize={0.2}
          color="white"
          anchorX="center"
        >
          {node.name}
        </Text>
      </group>
    </Float>
  );
}

interface MeshConnectionProps {
  from: [number, number, number];
  to: [number, number, number];
  index: number;
}

function MeshConnection({ from, to, index }: MeshConnectionProps) {
  const [opacity, setOpacity] = useState(0.4);

  useFrame((state) => {
    const time = state.clock.elapsedTime;
    setOpacity(0.3 + Math.sin(time * 2 + index) * 0.15);
  });

  const points = useMemo(() => {
    return [from, to] as [[number, number, number], [number, number, number]];
  }, [from, to]);

  return (
    <Line
      points={points}
      color={CYAN}
      lineWidth={1}
      transparent
      opacity={opacity}
    />
  );
}

interface DataPacketProps {
  from: [number, number, number];
  to: [number, number, number];
  delay: number;
}

function DataPacket({ from, to, delay }: DataPacketProps) {
  const meshRef = useRef<THREE.Mesh>(null);

  useFrame((state) => {
    if (!meshRef.current) return;
    const time = state.clock.elapsedTime;

    // Animate packet along the connection
    const t = ((time + delay) % 2) / 2; // 0 to 1 over 2 seconds

    meshRef.current.position.set(
      from[0] + (to[0] - from[0]) * t,
      from[1] + (to[1] - from[1]) * t,
      from[2] + (to[2] - from[2]) * t
    );

    // Fade in and out
    const material = meshRef.current.material as THREE.MeshBasicMaterial;
    const fadeT = Math.sin(t * Math.PI);
    material.opacity = fadeT * 0.8;

    // Scale based on position
    meshRef.current.scale.setScalar(0.08 + fadeT * 0.04);
  });

  return (
    <mesh ref={meshRef}>
      <sphereGeometry args={[1, 8, 8]} />
      <meshBasicMaterial color="white" transparent opacity={0.8} />
    </mesh>
  );
}
