// @a0n/gnosis/forkjoin-app/client/components/Relay3D/TopologyVisualization.tsx
/**
 * TopologyVisualization - 3D visualization of different network topologies
 *
 * Shows Star (signaling), Mesh (P2P), and Party Line (broadcast) patterns.
 */

import { useRef, useMemo, useState } from '@a0n/raect';
import { Canvas, useFrame } from '@a0n/aeon-3d/fiber';
import { Sphere, Float, Html, Line } from '@a0n/aeon-3d/drei';
import * as THREE from '@a0n/aeon-3d/three';
import { SafeBloom } from '../3D/SafeEffects';

const CYAN = '#06b6d4';
const PURPLE = '#a855f7';
const GREEN = '#10b981';

type TopologyType = 'star' | 'mesh' | 'partyline';

interface TopologyVisualizationProps {
  topology?: TopologyType;
  height?: number;
}

/**
 * Renders the Topology Visualization view.
 */
export function TopologyVisualization({
  topology = 'star',
  height = 280,
}: TopologyVisualizationProps) {
  return (
    <div style={{ width: '100%', height: `${height}px` }}>
      <Canvas camera={{ position: [0, 0, 8], fov: 45 }}>
        <ambientLight intensity={0.3} />
        <pointLight position={[10, 10, 10]} intensity={0.5} />
        <pointLight position={[-10, -10, 10]} intensity={0.3} color={CYAN} />

        {topology === 'star' && <StarTopology />}
        {topology === 'mesh' && <MeshTopology />}
        {topology === 'partyline' && <PartyLineTopology />}

        <SafeBloom
          intensity={0.4}
          luminanceThreshold={0.2}
          luminanceSmoothing={0.9}
        />
      </Canvas>
    </div>
  );
}

/**
 * Interactive topology selector with all three visualizations
 */
export function TopologyShowcase() {
  const [active, setActive] = useState<TopologyType>('star');

  return (
    <div className="topology-showcase">
      <div className="topology-tabs">
        <button
          className={`topology-tab ${active === 'star' ? 'active' : ''}`}
          onClick={() => setActive('star')}
        >
          Star (Signaling)
        </button>
        <button
          className={`topology-tab ${active === 'mesh' ? 'active' : ''}`}
          onClick={() => setActive('mesh')}
        >
          Mesh (P2P)
        </button>
        <button
          className={`topology-tab ${active === 'partyline' ? 'active' : ''}`}
          onClick={() => setActive('partyline')}
        >
          Party Line
        </button>
      </div>

      <TopologyVisualization topology={active} height={300} />

      <div className="topology-description">
        {active === 'star' && (
          <p>
            <strong>WebRTC Signaling:</strong> Central relay coordinates peer
            discovery. Peers connect through the signaling server, then
            establish direct connections.
          </p>
        )}
        {active === 'mesh' && (
          <p>
            <strong>Full Mesh:</strong> Every peer connects to every other peer
            directly. Maximum redundancy, optimal for small groups (2-8 peers).
          </p>
        )}
        {active === 'partyline' && (
          <p>
            <strong>Party Line:</strong> Selective forwarding for large groups.
            One sender, many receivers. Ideal for broadcasts and screen sharing.
          </p>
        )}
      </div>

      <style>{`
        .topology-showcase {
          background: rgba(0, 0, 0, 0.3);
          border-radius: 12px;
          overflow: hidden;
        }

        .topology-tabs {
          display: flex;
          border-bottom: 1px solid rgba(6, 182, 212, 0.2);
        }

        .topology-tab {
          flex: 1;
          padding: 12px 16px;
          background: none;
          border: none;
          color: var(--color-text-muted);
          font-size: 0.85rem;
          cursor: pointer;
          transition: all 0.2s ease;
        }

        .topology-tab:hover {
          background: rgba(6, 182, 212, 0.1);
          color: var(--color-text);
        }

        .topology-tab.active {
          background: rgba(6, 182, 212, 0.15);
          color: #06b6d4;
          border-bottom: 2px solid #06b6d4;
        }

        .topology-description {
          padding: 16px;
          font-size: 0.85rem;
          color: var(--color-text-secondary);
          border-top: 1px solid rgba(6, 182, 212, 0.2);
        }

        .topology-description strong {
          color: #06b6d4;
        }

        .topology-description p {
          margin: 0;
        }
      `}</style>
    </div>
  );
}

/**
 * Star Topology - WebRTC Signaling Pattern
 */
function StarTopology() {
  const groupRef = useRef<THREE.Group>(null);

  // Central signaling server + 5 peers around it
  const peers = useMemo(() => {
    const count = 5;
    return Array.from({ length: count }, (_, i) => {
      const angle = (i / count) * Math.PI * 2;
      const radius = 2.5;
      return {
        position: [Math.cos(angle) * radius, Math.sin(angle) * radius, 0] as [
          number,
          number,
          number
        ],
        label: `Peer ${i + 1}`,
      };
    });
  }, []);

  useFrame((state) => {
    if (!groupRef.current) return;
    const time = state.clock.elapsedTime;
    groupRef.current.rotation.z = time * 0.1;
  });

  return (
    <group ref={groupRef}>
      {/* Central Signaling Server */}
      <Float speed={2} floatIntensity={0.1}>
        <Sphere args={[0.4, 32, 32]} position={[0, 0, 0]}>
          <meshStandardMaterial
            color={PURPLE}
            emissive={PURPLE}
            emissiveIntensity={0.6}
            metalness={0.5}
            roughness={0.3}
          />
        </Sphere>
        <Html position={[0, -0.7, 0]} center>
          <span
            style={{
              color: PURPLE,
              fontSize: '10px',
              fontWeight: 600,
              whiteSpace: 'nowrap',
            }}
          >
            Relay Server
          </span>
        </Html>
      </Float>

      {/* Peer nodes */}
      {peers.map((peer, i) => (
        <PeerNode
          key={i}
          position={peer.position}
          label={peer.label}
          index={i}
          color={CYAN}
        />
      ))}

      {/* Connections from center to peers */}
      {peers.map((peer, i) => (
        <SignalingConnection
          key={`sig-${i}`}
          from={[0, 0, 0]}
          to={peer.position}
          index={i}
        />
      ))}

      {/* Data packets from center */}
      {peers.map((peer, i) => (
        <DataPacket
          key={`packet-${i}`}
          from={[0, 0, 0]}
          to={peer.position}
          delay={i * 0.4}
          color={PURPLE}
        />
      ))}
    </group>
  );
}

/**
 * Mesh Topology - Full P2P mesh
 */
function MeshTopology() {
  const groupRef = useRef<THREE.Group>(null);

  // 4 peers in mesh
  const peers = useMemo(
    () => [
      { position: [-1.5, 1.5, 0] as [number, number, number], label: 'Peer A' },
      { position: [1.5, 1.5, 0] as [number, number, number], label: 'Peer B' },
      {
        position: [-1.5, -1.5, 0] as [number, number, number],
        label: 'Peer C',
      },
      { position: [1.5, -1.5, 0] as [number, number, number], label: 'Peer D' },
    ],
    []
  );

  // All-to-all connections
  const connections = useMemo(() => {
    const conns: [number, number][] = [];
    for (let i = 0; i < peers.length; i++) {
      for (let j = i + 1; j < peers.length; j++) {
        conns.push([i, j]);
      }
    }
    return conns;
  }, [peers.length]);

  useFrame((state) => {
    if (!groupRef.current) return;
    const time = state.clock.elapsedTime;
    groupRef.current.rotation.y = Math.sin(time * 0.15) * 0.15;
    groupRef.current.rotation.x = Math.sin(time * 0.1) * 0.05;
  });

  return (
    <group ref={groupRef}>
      {/* Peer nodes */}
      {peers.map((peer, i) => (
        <PeerNode
          key={i}
          position={peer.position}
          label={peer.label}
          index={i}
          color={CYAN}
        />
      ))}

      {/* Mesh connections */}
      {connections.map(([from, to], i) => (
        <MeshConnection
          key={`conn-${i}`}
          from={peers[from].position}
          to={peers[to].position}
          index={i}
        />
      ))}

      {/* Data packets flowing bidirectionally */}
      {connections.map(([from, to], i) => (
        <DataPacket
          key={`packet-${i}`}
          from={peers[from].position}
          to={peers[to].position}
          delay={i * 0.3}
          color={CYAN}
        />
      ))}
    </group>
  );
}

/**
 * Party Line Topology - Selective Forwarding (SFU pattern)
 */
function PartyLineTopology() {
  const groupRef = useRef<THREE.Group>(null);

  // 1 sender, many receivers arranged in arc
  const sender = {
    position: [-2.5, 0, 0] as [number, number, number],
    label: 'Sender',
  };

  const receivers = useMemo(() => {
    const count = 6;
    return Array.from({ length: count }, (_, i) => {
      const angle = ((i - (count - 1) / 2) / count) * Math.PI * 0.8;
      return {
        position: [1.5 + Math.cos(angle) * 1.5, Math.sin(angle) * 2.5, 0] as [
          number,
          number,
          number
        ],
        label: `R${i + 1}`,
      };
    });
  }, []);

  useFrame((state) => {
    if (!groupRef.current) return;
    const time = state.clock.elapsedTime;
    groupRef.current.rotation.y = Math.sin(time * 0.1) * 0.1;
  });

  return (
    <group ref={groupRef}>
      {/* Sender node (larger) */}
      <Float speed={2} floatIntensity={0.1}>
        <Sphere args={[0.35, 32, 32]} position={sender.position}>
          <meshStandardMaterial
            color={GREEN}
            emissive={GREEN}
            emissiveIntensity={0.6}
            metalness={0.5}
            roughness={0.3}
          />
        </Sphere>
        <Html
          position={[sender.position[0], sender.position[1] - 0.6, 0]}
          center
        >
          <span
            style={{
              color: GREEN,
              fontSize: '10px',
              fontWeight: 600,
              whiteSpace: 'nowrap',
            }}
          >
            Sender
          </span>
        </Html>
      </Float>

      {/* Receiver nodes */}
      {receivers.map((receiver, i) => (
        <PeerNode
          key={i}
          position={receiver.position}
          label={receiver.label}
          index={i}
          color={CYAN}
          size={0.2}
        />
      ))}

      {/* Broadcast connections */}
      {receivers.map((receiver, i) => (
        <BroadcastConnection
          key={`bc-${i}`}
          from={sender.position}
          to={receiver.position}
          index={i}
        />
      ))}

      {/* Data packets broadcasting */}
      {receivers.map((receiver, i) => (
        <DataPacket
          key={`packet-${i}`}
          from={sender.position}
          to={receiver.position}
          delay={i * 0.15}
          color={GREEN}
        />
      ))}
    </group>
  );
}

// ===== Shared Components =====

interface PeerNodeProps {
  position: [number, number, number];
  label: string;
  index: number;
  color: string;
  size?: number;
}

function PeerNode({
  position,
  label,
  index,
  color,
  size = 0.25,
}: PeerNodeProps) {
  const meshRef = useRef<THREE.Mesh>(null);

  useFrame((state) => {
    if (!meshRef.current) return;
    const time = state.clock.elapsedTime;
    const scale = size + Math.sin(time * 2 + index) * 0.03;
    meshRef.current.scale.setScalar(scale / size);
  });

  return (
    <Float speed={2} floatIntensity={0.1}>
      <Sphere ref={meshRef} args={[size, 32, 32]} position={position}>
        <meshStandardMaterial
          color={color}
          emissive={color}
          emissiveIntensity={0.5}
          metalness={0.5}
          roughness={0.3}
        />
      </Sphere>
      <Html position={[position[0], position[1] - 0.45, position[2]]} center>
        <span
          style={{
            color: 'white',
            fontSize: '9px',
            opacity: 0.8,
            whiteSpace: 'nowrap',
          }}
        >
          {label}
        </span>
      </Html>
    </Float>
  );
}

interface ConnectionProps {
  from: [number, number, number];
  to: [number, number, number];
  index: number;
}

function SignalingConnection({ from, to, index }: ConnectionProps) {
  const [opacity, setOpacity] = useState(0.3);

  useFrame((state) => {
    const time = state.clock.elapsedTime;
    setOpacity(0.2 + Math.sin(time * 2 + index) * 0.1);
  });

  const points = useMemo(() => {
    return [from, to] as [[number, number, number], [number, number, number]];
  }, [from, to]);

  return (
    <Line
      points={points}
      color={PURPLE}
      lineWidth={1}
      transparent
      opacity={opacity}
    />
  );
}

function MeshConnection({ from, to, index }: ConnectionProps) {
  const [opacity, setOpacity] = useState(0.4);

  useFrame((state) => {
    const time = state.clock.elapsedTime;
    setOpacity(0.3 + Math.sin(time * 2 + index) * 0.15);
  });

  const points = useMemo(() => {
    return [from, to] as [[number, number, number], [number, number, number]];
  }, [from, to]);

  return (
    <Line
      points={points}
      color={CYAN}
      lineWidth={1}
      transparent
      opacity={opacity}
    />
  );
}

function BroadcastConnection({ from, to, index }: ConnectionProps) {
  const [opacity, setOpacity] = useState(0.35);

  useFrame((state) => {
    const time = state.clock.elapsedTime;
    setOpacity(0.25 + Math.sin(time * 3 + index * 0.5) * 0.15);
  });

  const points = useMemo(() => {
    return [from, to] as [[number, number, number], [number, number, number]];
  }, [from, to]);

  return (
    <Line
      points={points}
      color={GREEN}
      lineWidth={1}
      transparent
      opacity={opacity}
    />
  );
}

interface DataPacketProps {
  from: [number, number, number];
  to: [number, number, number];
  delay: number;
  color: string;
}

function DataPacket({ from, to, delay, color }: DataPacketProps) {
  const meshRef = useRef<THREE.Mesh>(null);

  useFrame((state) => {
    if (!meshRef.current) return;
    const time = state.clock.elapsedTime;

    // Animate packet along the connection
    const t = ((time + delay) % 1.5) / 1.5;

    meshRef.current.position.set(
      from[0] + (to[0] - from[0]) * t,
      from[1] + (to[1] - from[1]) * t,
      from[2] + (to[2] - from[2]) * t
    );

    // Fade in and out
    const material = meshRef.current.material as THREE.MeshBasicMaterial;
    const fadeT = Math.sin(t * Math.PI);
    material.opacity = fadeT * 0.9;

    // Scale based on position
    meshRef.current.scale.setScalar(0.06 + fadeT * 0.03);
  });

  return (
    <mesh ref={meshRef}>
      <sphereGeometry args={[1, 8, 8]} />
      <meshBasicMaterial color={color} transparent opacity={0.8} />
    </mesh>
  );
}
