// @a0n/gnosis/forkjoin-app/client/components/Flux3D/ArchitectureVisualization.tsx
/**
 * ArchitectureVisualization - 3D visualization of Aeon-Flux architecture
 *
 * Shows the Cloudflare Edge -> Router -> ESI -> Browser flow.
 */

import { useRef, useMemo } from '@a0n/raect';
import { Canvas, useFrame } from '@a0n/aeon-3d/fiber';
import { Text, RoundedBox, Float } from '@a0n/aeon-3d/drei';
import * as THREE from '@a0n/aeon-3d/three';
import { SafeBloom } from '../3D/SafeEffects';

const EMERALD = '#10b981';
const TEAL = '#14b8a6';

/**
 * Renders the Architecture Visualization view.
 */
export function ArchitectureVisualization() {
  return (
    <div style={{ width: '100%', height: '350px' }}>
      <Canvas camera={{ position: [0, 0, 10], fov: 45 }}>
        <ambientLight intensity={0.4} />
        <pointLight position={[10, 10, 10]} intensity={0.5} />
        <pointLight position={[-10, 5, 5]} intensity={0.3} color={EMERALD} />

        <ArchScene />

        <SafeBloom
          intensity={0.4}
          luminanceThreshold={0.3}
          luminanceSmoothing={0.9}
        />
      </Canvas>
    </div>
  );
}

function ArchScene() {
  const groupRef = useRef<THREE.Group>(null);

  useFrame((state) => {
    if (!groupRef.current) return;
    const time = state.clock.elapsedTime;
    groupRef.current.rotation.y = Math.sin(time * 0.1) * 0.1;
  });

  return (
    <group ref={groupRef}>
      {/* Cloudflare Edge Layer */}
      <EdgeLayer />

      {/* Router Layer */}
      <RouterLayer />

      {/* ESI Processor */}
      <ESILayer />

      {/* Browser */}
      <BrowserLayer />

      {/* Connecting data flows */}
      <DataFlows />
    </group>
  );
}

function EdgeLayer() {
  const services = [
    { name: 'WASM', x: -2.5 },
    { name: 'KV', x: -0.8 },
    { name: 'D1', x: 0.8 },
    { name: 'DO', x: 2.5 },
  ];

  return (
    <group position={[0, 2.5, 0]}>
      <Text
        position={[0, 1, 0]}
        fontSize={0.2}
        color="#808080"
        anchorX="center"
      >
        Cloudflare Edge
      </Text>

      {/* Edge container */}
      <RoundedBox args={[7, 1.5, 0.1]} radius={0.1} smoothness={4}>
        <meshStandardMaterial
          color="#1a1a2e"
          emissive={EMERALD}
          emissiveIntensity={0.05}
          transparent
          opacity={0.8}
        />
      </RoundedBox>

      {/* Service boxes */}
      {services.map((service, i) => (
        <Float key={service.name} speed={2} floatIntensity={0.1}>
          <group position={[service.x, 0, 0.1]}>
            <RoundedBox args={[1.2, 0.8, 0.05]} radius={0.05}>
              <meshStandardMaterial
                color={EMERALD}
                emissive={EMERALD}
                emissiveIntensity={0.3}
                transparent
                opacity={0.9}
              />
            </RoundedBox>
            <Text
              position={[0, 0, 0.08]}
              fontSize={0.18}
              color="white"
              anchorX="center"
            >
              {service.name}
            </Text>
          </group>
        </Float>
      ))}
    </group>
  );
}

function RouterLayer() {
  return (
    <Float speed={1.5} floatIntensity={0.1}>
      <group position={[0, 0.5, 0]}>
        <RoundedBox args={[4, 0.7, 0.1]} radius={0.1}>
          <meshStandardMaterial
            color={TEAL}
            emissive={TEAL}
            emissiveIntensity={0.3}
            transparent
            opacity={0.9}
          />
        </RoundedBox>
        <Text
          position={[0, 0, 0.1]}
          fontSize={0.18}
          color="white"
          anchorX="center"
        >
          Aeon-Flux Router (Personalization)
        </Text>
      </group>
    </Float>
  );
}

function ESILayer() {
  return (
    <Float speed={1.5} floatIntensity={0.1}>
      <group position={[0, -0.8, 0]}>
        <RoundedBox args={[3.5, 0.7, 0.1]} radius={0.1}>
          <meshStandardMaterial
            color="#336699"
            emissive="#336699"
            emissiveIntensity={0.3}
            transparent
            opacity={0.9}
          />
        </RoundedBox>
        <Text
          position={[0, 0, 0.1]}
          fontSize={0.18}
          color="white"
          anchorX="center"
        >
          ESI Processor (AI at render)
        </Text>
      </group>
    </Float>
  );
}

function BrowserLayer() {
  return (
    <group position={[0, -2.5, 0]}>
      <Float speed={2} floatIntensity={0.15}>
        <RoundedBox args={[2.5, 1, 0.1]} radius={0.1}>
          <meshStandardMaterial
            color="#f59e0b"
            emissive="#f59e0b"
            emissiveIntensity={0.3}
            transparent
            opacity={0.9}
          />
        </RoundedBox>
        <Text
          position={[0, 0.2, 0.1]}
          fontSize={0.18}
          color="white"
          anchorX="center"
        >
          Browser (React)
        </Text>
        <Text
          position={[0, -0.15, 0.1]}
          fontSize={0.1}
          color="#999999"
          anchorX="center"
        >
          180KB hydration | Instant nav
        </Text>
      </Float>
    </group>
  );
}

function DataFlows() {
  const flowsRef = useRef<THREE.Group>(null);

  useFrame((state) => {
    if (!flowsRef.current) return;
    const time = state.clock.elapsedTime;

    // Animate flow opacity
    flowsRef.current.children.forEach((child, i) => {
      const mesh = child as THREE.Mesh;
      const material = mesh.material as THREE.MeshBasicMaterial;
      material.opacity = 0.3 + Math.sin(time * 3 + i * 2) * 0.2;
    });
  });

  return (
    <group ref={flowsRef}>
      {/* Edge to Router */}
      <FlowArrow start={[0, 1.6, 0]} end={[0, 0.95, 0]} color={EMERALD} />

      {/* Router to ESI */}
      <FlowArrow start={[0, 0.1, 0]} end={[0, -0.4, 0]} color={TEAL} />

      {/* ESI to Browser */}
      <FlowArrow start={[0, -1.2, 0]} end={[0, -1.9, 0]} color="#336699" />
    </group>
  );
}

interface FlowArrowProps {
  start: [number, number, number];
  end: [number, number, number];
  color: string;
}

function FlowArrow({ start, end, color }: FlowArrowProps) {
  const meshRef = useRef<THREE.Mesh>(null);

  useFrame((state) => {
    if (!meshRef.current) return;
    const time = state.clock.elapsedTime;
    const material = meshRef.current.material as THREE.MeshBasicMaterial;
    material.opacity = 0.4 + Math.sin(time * 4) * 0.2;
  });

  const length = Math.sqrt(
    (end[0] - start[0]) ** 2 +
      (end[1] - start[1]) ** 2 +
      (end[2] - start[2]) ** 2
  );

  const midX = (start[0] + end[0]) / 2;
  const midY = (start[1] + end[1]) / 2;
  const midZ = (start[2] + end[2]) / 2;

  return (
    <mesh ref={meshRef} position={[midX, midY, midZ]}>
      <cylinderGeometry args={[0.02, 0.02, length, 8]} />
      <meshBasicMaterial color={color} transparent opacity={0.5} />
    </mesh>
  );
}
