// @a0n/.ignored_gnosis/distributed-inference-host/src/__tests__/http-byte-parity.test.ts
/**
 * HTTP byte-parity test for the Pneuma fat-station residual transport.
 *
 * Hypothesis
 * ----------
 * Every kernel parity test in the Pneuma debug session ran in-process
 * (Vec<f32> handed directly between pipelines), but the deployed mesh ships
 * residuals through HTTP body bytes. The mesh emits gibberish even when the
 * kernel is bit-perfect. This test sends a known f32 buffer to a minimal
 * /echo route on `fat-station-echo` and verifies bit-equality round-trip.
 *
 * If PASS for all sizes: the HTTP body codec is innocent and the gibberish
 * source lives downstream (likely the chunked-attn-chunk path that did not
 * receive the Wallington Rotation).
 *
 * If FAIL: surface the corruption pattern (endianness flip? padding?
 * content-encoding? truncation at a power-of-two boundary?).
 *
 * Helpers
 * -------
 * `floatArrayToBody` and `readFloatArrayResponse` are copied verbatim from
 * `src/station.ts` so this test exercises the EXACT code path station.ts
 * uses. The constraint set says "do not modify station.ts" so we cannot
 * promote them to exports — copy is documented and intentional.
 *
 * Server
 * ------
 * Started by `scripts/run-byte-parity.sh` on 127.0.0.1:9999.
 *
 * Reference: docs/pneuma/http-byte-parity.md
 */

import { afterAll, describe, expect, it } from 'vitest';

const BASE_URL = process.env.ECHO_BASE_URL ?? 'http://127.0.0.1:9999';

// Probe the echo server once at module load. When unreachable (the default
// outside `scripts/run-byte-parity.sh`), skip the suite instead of failing
// the whole package's `:test` target. The fat-station-echo binary is a
// separate Rust integration harness owned by the Pneuma debug workflow.
async function probeEchoServer(): Promise<boolean> {
  const ctrl = new AbortController();
  const timer = setTimeout(() => ctrl.abort(), 750);
  try {
    const res = await fetch(`${BASE_URL}/health`, { signal: ctrl.signal });
    return res.ok;
  } catch {
    return false;
  } finally {
    clearTimeout(timer);
  }
}

const echoReachable = await probeEchoServer();
const describeIfEcho = echoReachable ? describe : describe.skip;

// ── helpers (verbatim copy of src/station.ts:592 + :604) ─────────────────

/** Copy a Float32Array view into a fresh ArrayBuffer suitable for Request.body. */
function floatArrayToBody(arr: Float32Array): ArrayBuffer {
  const buf = new ArrayBuffer(arr.byteLength);
  new Uint8Array(buf).set(new Uint8Array(arr.buffer, arr.byteOffset, arr.byteLength));
  return buf;
}

async function readFloatArrayResponse(res: Response, label: string): Promise<Float32Array> {
  if (!res.ok) {
    let detail = '';
    try {
      detail = await res.text();
    } catch {
      // ignore
    }
    throw new Error(
      `Station ${label}: HTTP ${res.status} ${res.statusText}${detail ? ` -- ${detail}` : ''}`,
    );
  }
  const buf = await res.arrayBuffer();
  if (buf.byteLength % 4 !== 0) {
    throw new Error(
      `Station ${label}: response byteLength ${buf.byteLength} is not a multiple of 4 (Float32)`,
    );
  }
  return new Float32Array(buf);
}

// ── seeded RNG (mulberry32) so payloads are reproducible across runs ─────

function mulberry32(seed: number): () => number {
  let s = seed >>> 0;
  return () => {
    s = (s + 0x6d2b79f5) >>> 0;
    let t = s;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

/** Build a Float32Array of `nFloats` values in [-1, 1) from `seed`. */
function makeSeededFloats(nFloats: number, seed: number): Float32Array {
  const rng = mulberry32(seed);
  const out = new Float32Array(nFloats);
  for (let i = 0; i < nFloats; i++) {
    out[i] = (rng() * 2 - 1) as number;
  }
  return out;
}

// ── transport ─────────────────────────────────────────────────────────────

async function echo(input: Float32Array, label: string): Promise<Float32Array> {
  const body = floatArrayToBody(input);
  const res = await fetch(`${BASE_URL}/echo`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/octet-stream',
      'X-Request-Id': `parity-${label}`,
    },
    body,
  });
  return await readFloatArrayResponse(res, '/echo');
}

/** Bit-equal compare via Uint8 view (handles NaN, denormals, ±0). */
function bytesEqual(a: Float32Array, b: Float32Array): { ok: boolean; firstDelta?: number } {
  if (a.byteLength !== b.byteLength) return { ok: false, firstDelta: -1 };
  const ua = new Uint8Array(a.buffer, a.byteOffset, a.byteLength);
  const ub = new Uint8Array(b.buffer, b.byteOffset, b.byteLength);
  for (let i = 0; i < ua.length; i++) {
    if (ua[i] !== ub[i]) return { ok: false, firstDelta: i };
  }
  return { ok: true };
}

// ── boot check ────────────────────────────────────────────────────────────

afterAll(() => {
  // Server lifecycle is owned by run-byte-parity.sh; nothing to tear down here.
});

// ── test cases ────────────────────────────────────────────────────────────

const SIZES_BYTES = [
  { label: 'single-token (14336B)', bytes: 14336 },
  { label: '4-token chunk (57344B)', bytes: 57344 },
  { label: '15-token chunk (215040B)', bytes: 215040 },
  { label: 'vocab-size logits (593760B)', bytes: 593760 },
];

describeIfEcho('HTTP residual transport byte parity (fat-station-echo)', () => {
  for (const { label, bytes } of SIZES_BYTES) {
    it(`round-trips ${label} bit-perfectly (seed=42)`, async () => {
      expect(bytes % 4).toBe(0);
      const nFloats = bytes / 4;
      const input = makeSeededFloats(nFloats, 42);
      const echoed = await echo(input, `seed42-${bytes}`);
      expect(echoed.byteLength).toBe(input.byteLength);
      expect(echoed.length).toBe(input.length);
      const cmp = bytesEqual(input, echoed);
      if (!cmp.ok) {
        const i = cmp.firstDelta!;
        throw new Error(
          `byte mismatch at offset ${i}: ` +
            `input[${Math.floor(i / 4)}]=${input[Math.floor(i / 4)]} ` +
            `echo[${Math.floor(i / 4)}]=${echoed[Math.floor(i / 4)]}`,
        );
      }
      expect(cmp.ok).toBe(true);
    }, 30_000);
  }

  it('round-trips edge-case f32 values (NaN, Infinity, denormals, zeros) bit-perfectly', async () => {
    // Hand-crafted edge values. Use Uint32Array → Float32Array view so we
    // control the exact bit pattern (Float32Array literal cannot express
    // all NaN payloads).
    const u32 = new Uint32Array([
      0x00000000, // +0.0
      0x80000000, // -0.0
      0x3f800000, // 1.0
      0xbf800000, // -1.0
      0x7f800000, // +Infinity
      0xff800000, // -Infinity
      0x7fc00000, // canonical quiet NaN
      0x7fa00000, // signaling NaN payload
      0xffc00012, // negative NaN with payload
      0x00000001, // smallest positive denormal
      0x007fffff, // largest denormal
      0x00800000, // smallest positive normal
      0x7f7fffff, // largest finite
      0x80000001, // smallest negative denormal
      0x33800000, // tiny positive
      0xb3800000, // tiny negative
    ]);
    const input = new Float32Array(u32.buffer.slice(0));
    const echoed = await echo(input, 'edge');
    expect(echoed.byteLength).toBe(input.byteLength);

    // Compare via Uint32 view — Float32Array .every comparisons treat NaN
    // as not-equal, so we MUST go through the bit pattern.
    const u32In = new Uint32Array(input.buffer);
    const u32Out = new Uint32Array(echoed.buffer);
    for (let i = 0; i < u32In.length; i++) {
      if (u32In[i] !== u32Out[i]) {
        throw new Error(
          `bit mismatch at index ${i}: ` +
            `in=0x${u32In[i]!.toString(16).padStart(8, '0')} ` +
            `out=0x${u32Out[i]!.toString(16).padStart(8, '0')}`,
        );
      }
    }
  }, 30_000);
});
