// @a0n/gnosis/src/capabilities/profiles.ts
import type {
  CapabilityProfile,
  HostCapability,
  RuntimeTarget,
} from './types';

const NODE_CAPABILITIES: HostCapability[] = [
  'net.tcp.client',
  'net.tcp.server',
  'net.udp',
  'fs.local',
  'fs.durable',
  'auth.ucan',
  'auth.zk',
  'auth.custodial',
  'process.spawn',
  'process.docker',
];

const BUN_CAPABILITIES: HostCapability[] = [
  'net.tcp.client',
  'net.tcp.server',
  'net.udp',
  'fs.local',
  'fs.durable',
  'auth.ucan',
  'auth.zk',
  'auth.custodial',
  'process.spawn',
  'process.docker',
];

const WORKERS_CAPABILITIES: HostCapability[] = [
  'net.tcp.client',
  'fs.local',
  'auth.ucan',
  'auth.zk',
  'auth.custodial',
];

const TAURI_EDGE_CAPABILITIES: HostCapability[] = [
  'fs.local',
  'fs.durable',
  'auth.ucan',
  'auth.zk',
];

const WASM_BROWSER_CAPABILITIES: HostCapability[] = ['auth.ucan', 'auth.zk'];

const WASI_CAPABILITIES: HostCapability[] = [
  'fs.local',
  'auth.ucan',
  'auth.zk',
];

export const CAPABILITY_PROFILES: Record<
  Exclude<RuntimeTarget, 'agnostic'>,
  CapabilityProfile
> = {
  node: {
    target: 'node',
    supported: new Set(NODE_CAPABILITIES),
    description:
      'Node.js host with TCP server/client, UDP, and durable local file access.',
  },
  bun: {
    target: 'bun',
    supported: new Set(BUN_CAPABILITIES),
    description:
      'Bun host with Node-compatible networking, durable local file access, subprocess spawning, and Docker CLI access.',
  },
  gnode: {
    target: 'gnode',
    supported: new Set(NODE_CAPABILITIES),
    description:
      'gnode host with Node-compatible networking, local file access, and topology compilation.',
  },
  workers: {
    target: 'workers',
    supported: new Set(WORKERS_CAPABILITIES),
    description:
      'Cloudflare Workers host with outbound TCP client support only (no TCP server, no UDP, no durable local disk).',
  },
  'tauri-edge': {
    target: 'tauri-edge',
    supported: new Set(TAURI_EDGE_CAPABILITIES),
    description:
      'Semantic Tauri edge runtime on the Gnosis WASM lane with local spool hydration, deterministic D1 reconstruction, and UCAN/ZK admission.',
  },
  'wasm-browser': {
    target: 'wasm-browser',
    supported: new Set(WASM_BROWSER_CAPABILITIES),
    description:
      'Browser-hosted WASM guest with compute-only auth support and no raw TCP, UDP, or durable local disk.',
  },
  wasi: {
    target: 'wasi',
    supported: new Set(WASI_CAPABILITIES),
    description:
      'WASI guest with local filesystem access and compute/auth support, but no raw TCP server/client or UDP.',
  },
};
