// @a0n/gnosis/src/capabilities/validate.ts
import { CAPABILITY_PROFILES } from './profiles';
import type {
  CapabilityIssue,
  CapabilityRequirement,
  CapabilityValidationReport,
  HostCapability,
  RuntimeTarget,
} from './types';

function uniqueCapabilities(
  requirements: readonly CapabilityRequirement[]
): HostCapability[] {
  return [
    ...new Set(requirements.map((requirement) => requirement.capability)),
  ].sort();
}

function appendDeclarationContractIssues(
  requirements: readonly CapabilityRequirement[],
  issues: CapabilityIssue[],
  target: RuntimeTarget
): void {
  const requirementsByNode = new Map<string, CapabilityRequirement[]>();

  for (const requirement of requirements) {
    const existing = requirementsByNode.get(requirement.nodeId) ?? [];
    existing.push(requirement);
    requirementsByNode.set(requirement.nodeId, existing);
  }

  for (const [nodeId, nodeRequirements] of requirementsByNode.entries()) {
    const declared = new Set(
      nodeRequirements
        .filter((requirement) => requirement.source === 'declaration')
        .map((requirement) => requirement.capability)
    );
    if (declared.size === 0) {
      continue;
    }

    const inferred = new Set(
      nodeRequirements
        .filter((requirement) => requirement.source === 'inference')
        .map((requirement) => requirement.capability)
    );

    for (const capability of inferred) {
      if (!declared.has(capability)) {
        issues.push({
          capability,
          severity: 'error',
          target,
          nodeId,
          message: `Node ${nodeId} inferred ${capability} without declaring it in effects/requires.`,
        });
      }
    }

    for (const capability of declared) {
      if (!inferred.has(capability)) {
        issues.push({
          capability,
          severity: 'warning',
          target,
          nodeId,
          message: `Node ${nodeId} declares ${capability} but no matching usage was inferred.`,
        });
      }
    }
  }
}

/**
 * Handles the gnosis validate Capabilities For Target workflow.
 */
export function validateCapabilitiesForTarget(
  requirements: readonly CapabilityRequirement[],
  target: RuntimeTarget
): CapabilityValidationReport {
  const requiredUnique = uniqueCapabilities(requirements);
  const issues: CapabilityIssue[] = [];
  appendDeclarationContractIssues(requirements, issues, target);

  if (target === 'agnostic') {
    return {
      target,
      required: [...requirements],
      requiredUnique,
      issues,
      ok: !issues.some((issue) => issue.severity === 'error'),
    };
  }

  const profile = CAPABILITY_PROFILES[target];

  for (const capability of requiredUnique) {
    if (!profile.supported.has(capability)) {
      issues.push({
        capability,
        severity: 'error',
        target,
        message: `${capability} is not supported on ${target}.`,
      });
      continue;
    }

    if (
      (target === 'workers' || target === 'tauri-edge') &&
      capability === 'fs.local'
    ) {
      issues.push({
        capability,
        severity: 'warning',
        target,
        message: `${target} uses fs.local as a host-local spool/cache lane; it should not be treated as durable authority across restarts.`,
      });
    }
  }

  return {
    target,
    required: [...requirements],
    requiredUnique,
    issues,
    ok: !issues.some((issue) => issue.severity === 'error'),
  };
}
