// @a0n/gnosis/src/__tests__/arena-merge-gate.test.ts
import { Verdict } from '@a0n/bitwise';

import { evaluateArenaMergeGate } from '../arena-merge-gate';

const baseBundle = {
  schemaVersion: '1.0.0' as const,
  patch_content_id: 'deadbeefdeadbeefdeadbeefdeadbeef',
  toolchain_id: 'toolchain_pin_ok_minimum',
  ci_config_id: 'ci_config_hash_ok_minimum',
  gates: [{ name: 'lake_build', exit: 0, log_hash: 'loghash_ok_minimum_len' }],
  formal: { sorry_count_delta: 0 },
};

export default function law(): void {
  const pass = evaluateArenaMergeGate({ verificationBundle: baseBundle });
  if (!pass.ok) {
    throw new Error(`expected ok, got ${JSON.stringify(pass)}`);
  }

  const bad = evaluateArenaMergeGate({
    verificationBundle: { ...baseBundle, patch_content_id: 'short' },
  });
  if (bad.ok || !bad.reasons.some((r) => r.startsWith('verification_'))) {
    throw new Error(`expected verification failure, got ${JSON.stringify(bad)}`);
  }

  const cool = evaluateArenaMergeGate({
    verificationBundle: baseBundle,
    cooling: { state: { arousalProxy: 0, alertRatePerMin: 0, mode: 'cooling' } },
  });
  if (cool.ok || !cool.reasons.includes('cooling_active')) {
    throw new Error(`expected cooling block, got ${JSON.stringify(cool)}`);
  }

  // New quarantine/defer path: cooling alone is a SOFT hold (Abstain), not a
  // hard reject — the merge is sound, the system is just too hot right now.
  if (cool.ok || cool.verdict !== Verdict.Abstain) {
    throw new Error(`expected cooling to quarantine (Abstain), got ${JSON.stringify(cool)}`);
  }
  // Clean merge admits (Accept); hard verification failure declines (Decline).
  if (!pass.ok || pass.verdict !== Verdict.Accept) {
    throw new Error(`expected clean merge to admit (Accept), got ${JSON.stringify(pass)}`);
  }
  if (bad.ok || bad.verdict !== Verdict.Decline) {
    throw new Error(`expected verification failure to decline (Decline), got ${JSON.stringify(bad)}`);
  }

  const rollback = evaluateArenaMergeGate({
    verificationBundle: baseBundle,
    rollbackReplay: {
      policy: {
        lambda2Star: 1,
        tauWindows: 1,
        integralBudgetW: 0,
        beta1Min: 5,
      },
      samples: [{ lambda2Hat: 0.1, beta1Hat: 10, t: 0 }],
    },
  });
  if (
    rollback.ok ||
    !rollback.reasons.some((r) => r === 'rollback_consecutive_low' || r === 'rollback_integral_deficit')
  ) {
    throw new Error(`expected rollback block, got ${JSON.stringify(rollback)}`);
  }
}
