// @a0n/gnosis/src/__tests__/betti-wasm-store.test.ts
/**
 * Betti → WASM registry store tests.
 *
 * Covers the fail-closed validation/integrity core (sha256 over
 * `bytes[..byteLength]`, EXCLUSIVE of the trailing custom section), the D1
 * upsert + R2 object pair against minimal structural fakes, the warning
 * flags when handles are absent, load-side re-verification, and the local
 * file-registry mode's deterministic roundtrip.
 */
import { createHash } from 'node:crypto';
import { mkdtemp, readFile, rm } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { describe, expect, test } from '@a0n/gnosis/test';
import type { BettiWasmManifest } from '../betti-wasm-manifest';
import {
  bettiWasmManifestKey,
  bettiWasmModuleKey,
  ensureBettiWasmArtifactsTable,
  loadBettiWasmArtifact,
  storeBettiWasmArtifact,
  storeBettiWasmArtifactToDirectory,
  validateBettiWasmArtifactPair,
  type BettiWasmArtifactRow,
  type BettiWasmR2Bucket,
  type D1Database,
  type D1PreparedStatement,
} from '../betti-wasm-store';
import { fragmentKey } from '../polyglot-vfs';

// Hand-assembled scalar-v1 module: (func (export "main") (param f64)
// (result f64) local.get 0 f64.const 2 f64.mul) — doubles its argument.
const SCALAR_DOUBLE_WASM = Uint8Array.from([
  0x00, 0x61, 0x73, 0x6d, 0x01, 0x00, 0x00, 0x00, // header
  0x01, 0x06, 0x01, 0x60, 0x01, 0x7c, 0x01, 0x7c, // type: (f64) -> f64
  0x03, 0x02, 0x01, 0x00, // function section
  0x07, 0x08, 0x01, 0x04, 0x6d, 0x61, 0x69, 0x6e, 0x00, 0x00, // export "main"
  0x0a, 0x10, 0x01, 0x0e, 0x00, 0x20, 0x00, 0x44, // code: local.get 0,
  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x40, // f64.const 2,
  0xa2, 0x0b, // f64.mul, end
]);

function leb128(value: number): number[] {
  const bytes: number[] = [];
  let remaining = value;
  do {
    let byte = remaining & 0x7f;
    remaining >>>= 7;
    if (remaining !== 0) {
      byte |= 0x80;
    }
    bytes.push(byte);
  } while (remaining !== 0);
  return bytes;
}

/** Append a wasm custom section (id 0) with the given name + payload. */
function appendCustomSection(
  core: Uint8Array,
  name: string,
  payload: string
): Uint8Array {
  const encoder = new TextEncoder();
  const nameBytes = encoder.encode(name);
  const payloadBytes = encoder.encode(payload);
  const body = [
    ...leb128(nameBytes.length),
    ...nameBytes,
    ...payloadBytes,
  ];
  const section = [0x00, ...leb128(body.length), ...body];
  const out = new Uint8Array(core.length + section.length);
  out.set(core, 0);
  out.set(Uint8Array.from(section), core.length);
  return out;
}

interface ArtifactPair {
  readonly wasmBytes: Uint8Array;
  readonly manifestJson: string;
  readonly manifest: BettiWasmManifest;
}

function makePair(overrides: {
  artifactId: string;
  filePath?: string;
  functionName?: string;
}): ArtifactPair {
  const core = SCALAR_DOUBLE_WASM;
  const manifest = {
    artifact_version: 'betti-wasm/1',
    artifactId: overrides.artifactId,
    topologyHash: 'aaaa5555aaaa5555',
    sourceHash: '5555aaaa5555aaaa',
    language: 'typescript',
    filePath: overrides.filePath ?? 'gnode/benchmarks/double.ts',
    functionName: overrides.functionName ?? 'main',
    certified_class: 'numeric-loop',
    tier: 'certified-direct',
    export_name: 'main',
    abi: 'scalar-v1',
    imports: [],
    pure: true,
    limits: { memoryPagesMax: 1 },
    byteLength: core.length,
    sha256: createHash('sha256').update(core).digest('hex'),
    decline_semantics: 'trap',
    wasm_opt: false,
  };
  const manifestJson = JSON.stringify(manifest, null, 2);
  // Mirror the emitter layout: manifest custom section trails the hashed bytes.
  const wasmBytes = appendCustomSection(
    core,
    'gnosis-betti-manifest',
    manifestJson
  );
  return {
    wasmBytes,
    manifestJson,
    manifest: manifest as unknown as BettiWasmManifest,
  };
}

/** Column set of the table BEFORE the Wave D3 migration (no owner/price). */
const PRE_D3_COLUMNS = [
  'artifact_id',
  'key',
  'file_path',
  'function_name',
  'language',
  'tier',
  'pure',
  'manifest_json',
  'r2_key',
  'byte_length',
  'sha256',
  'created_at',
] as const;

const CURRENT_COLUMNS = [...PRE_D3_COLUMNS, 'owner_did', 'price_cents'] as const;

/**
 * Minimal structural D1 fake covering the store's exact statements, with a
 * simulated schema (column set) so `CREATE TABLE IF NOT EXISTS` no-ops on an
 * existing table and `ALTER TABLE ... ADD COLUMN` behaves like SQLite
 * (throws `duplicate column name` when the column exists). Construct with
 * `PRE_D3_COLUMNS` to model a database created before the migration.
 */
class FakeD1 implements D1Database {
  readonly rows = new Map<string, BettiWasmArtifactRow>();
  readonly executed: string[] = [];
  readonly columns = new Set<string>();

  constructor(existingColumns: readonly string[] = []) {
    for (const column of existingColumns) {
      this.columns.add(column);
    }
  }

  prepare(query: string): D1PreparedStatement {
    const rows = this.rows;
    const executed = this.executed;
    const columns = this.columns;
    let bound: unknown[] = [];
    const statement: D1PreparedStatement = {
      bind(...values: unknown[]) {
        bound = values;
        return statement;
      },
      async run() {
        executed.push(query);
        if (query.startsWith('CREATE TABLE IF NOT EXISTS')) {
          // No-op when the table already exists (columns pre-seeded).
          if (columns.size === 0) {
            for (const column of CURRENT_COLUMNS) {
              columns.add(column);
            }
          }
          return;
        }
        if (query.startsWith('CREATE INDEX')) {
          return;
        }
        const alterMatch = query.match(
          /^ALTER TABLE \S+ ADD COLUMN (\S+)\s/
        );
        if (alterMatch) {
          const column = alterMatch[1];
          if (columns.has(column)) {
            throw new Error(`duplicate column name: ${column}`);
          }
          columns.add(column);
          return;
        }
        if (query.trimStart().startsWith('INSERT OR REPLACE')) {
          for (const column of ['owner_did', 'price_cents']) {
            if (query.includes(column) && !columns.has(column)) {
              throw new Error(`table has no column named ${column}`);
            }
          }
          const [
            artifact_id,
            key,
            file_path,
            function_name,
            language,
            tier,
            pure,
            manifest_json,
            r2_key,
            byte_length,
            sha256,
            created_at,
            owner_did,
            price_cents,
          ] = bound as [
            string,
            string,
            string,
            string,
            string,
            string,
            number,
            string,
            string | null,
            number,
            string,
            string,
            string | null,
            number | null,
          ];
          rows.set(artifact_id, {
            artifact_id,
            key,
            file_path,
            function_name,
            language,
            tier,
            pure,
            manifest_json,
            r2_key,
            byte_length,
            sha256,
            created_at,
            owner_did,
            price_cents,
          });
          return;
        }
        throw new Error(`FakeD1.run: unsupported statement: ${query}`);
      },
      async first<T>() {
        executed.push(query);
        if (query.includes('WHERE artifact_id = ?')) {
          return (rows.get(bound[0] as string) ?? null) as T | null;
        }
        if (query.includes('WHERE key = ?')) {
          const matches = [...rows.values()]
            .filter((row) => row.key === bound[0])
            .sort((left, right) =>
              right.created_at.localeCompare(left.created_at)
            );
          return (matches[0] ?? null) as T | null;
        }
        throw new Error(`FakeD1.first: unsupported statement: ${query}`);
      },
      async all<T>() {
        throw new Error(`FakeD1.all: unsupported statement: ${query}`);
        return { results: [] as T[] };
      },
    };
    return statement;
  }

  async batch<T>(): Promise<T[]> {
    throw new Error('FakeD1.batch: not supported');
  }

  async exec(): Promise<void> {
    throw new Error('FakeD1.exec: not supported');
  }
}

class FakeR2 implements BettiWasmR2Bucket {
  readonly objects = new Map<string, Uint8Array>();

  async put(key: string, value: Uint8Array | string): Promise<unknown> {
    this.objects.set(
      key,
      typeof value === 'string'
        ? new TextEncoder().encode(value)
        : Uint8Array.from(value)
    );
    return null;
  }

  async get(
    key: string
  ): Promise<{ arrayBuffer(): Promise<ArrayBuffer> } | null> {
    const stored = this.objects.get(key);
    if (!stored) {
      return null;
    }
    return {
      arrayBuffer: async () =>
        stored.buffer.slice(
          stored.byteOffset,
          stored.byteOffset + stored.byteLength
        ) as ArrayBuffer,
    };
  }
}

const FIXED_NOW = () => new Date('2026-07-03T00:00:00.000Z');

describe('betti wasm store — validation core', () => {
  test('a valid pair validates (custom section excluded from the hash)', async () => {
    const pair = makePair({ artifactId: 'aaaaaaaaaaaaaaaa' });
    expect(pair.wasmBytes.length).toBeGreaterThan(pair.manifest.byteLength);
    const manifest = await validateBettiWasmArtifactPair(
      pair.wasmBytes,
      pair.manifestJson
    );
    expect(manifest.artifactId).toBe('aaaaaaaaaaaaaaaa');
  });

  test('tampered module bytes fail closed', async () => {
    const pair = makePair({ artifactId: 'aaaaaaaaaaaaaaaa' });
    const tampered = Uint8Array.from(pair.wasmBytes);
    tampered[16] ^= 0xff; // inside [0, byteLength)
    let failure: unknown = null;
    try {
      await validateBettiWasmArtifactPair(tampered, pair.manifestJson);
    } catch (error) {
      failure = error;
    }
    expect(String(failure)).toContain('integrity failure');
  });

  test('a truncated module fails closed', async () => {
    const pair = makePair({ artifactId: 'aaaaaaaaaaaaaaaa' });
    const truncated = pair.wasmBytes.subarray(
      0,
      pair.manifest.byteLength - 4
    );
    let failure: unknown = null;
    try {
      await validateBettiWasmArtifactPair(truncated, pair.manifestJson);
    } catch (error) {
      failure = error;
    }
    expect(String(failure)).toContain('exceeds module size');
  });

  test('malformed manifest JSON and invalid manifests fail closed', async () => {
    const pair = makePair({ artifactId: 'aaaaaaaaaaaaaaaa' });
    let jsonFailure: unknown = null;
    try {
      await validateBettiWasmArtifactPair(pair.wasmBytes, '{nope');
    } catch (error) {
      jsonFailure = error;
    }
    expect(String(jsonFailure)).toContain('not valid JSON');

    const invalid = JSON.parse(pair.manifestJson) as Record<string, unknown>;
    invalid['artifactId'] = 'not-hex';
    let manifestFailure: unknown = null;
    try {
      await validateBettiWasmArtifactPair(
        pair.wasmBytes,
        JSON.stringify(invalid)
      );
    } catch (error) {
      manifestFailure = error;
    }
    expect(String(manifestFailure)).toContain('artifactId');
  });
});

describe('betti wasm store — D1 + R2', () => {
  test('stores the row and the R2 object pair', async () => {
    const pair = makePair({ artifactId: 'aaaaaaaaaaaaaaaa' });
    const db = new FakeD1();
    const r2 = new FakeR2();

    const result = await storeBettiWasmArtifact({
      db,
      r2,
      wasmBytes: pair.wasmBytes,
      manifestJson: pair.manifestJson,
      now: FIXED_NOW,
    });

    expect(result.storedToD1).toBe(true);
    expect(result.storedToR2).toBe(true);
    expect(result.warnings.length).toBe(0);
    expect(result.key).toBe(
      fragmentKey(pair.manifest.filePath, pair.manifest.functionName)
    );

    const row = db.rows.get('aaaaaaaaaaaaaaaa');
    expect(row).not.toBe(undefined);
    expect(row?.key).toBe(result.key);
    expect(row?.pure).toBe(1);
    expect(row?.tier).toBe('certified-direct');
    expect(row?.byte_length).toBe(pair.manifest.byteLength);
    expect(row?.sha256).toBe(pair.manifest.sha256);
    expect(row?.r2_key).toBe(bettiWasmModuleKey('aaaaaaaaaaaaaaaa'));
    expect(row?.created_at).toBe('2026-07-03T00:00:00.000Z');

    expect(
      r2.objects.get(bettiWasmModuleKey('aaaaaaaaaaaaaaaa'))?.length
    ).toBe(pair.wasmBytes.length);
    expect(
      new TextDecoder().decode(
        r2.objects.get(bettiWasmManifestKey('aaaaaaaaaaaaaaaa'))
      )
    ).toBe(pair.manifestJson);
  });

  test('missing R2 skips the upload with a warning and a NULL r2_key', async () => {
    const pair = makePair({ artifactId: 'aaaaaaaaaaaaaaaa' });
    const db = new FakeD1();
    const result = await storeBettiWasmArtifact({
      db,
      wasmBytes: pair.wasmBytes,
      manifestJson: pair.manifestJson,
      now: FIXED_NOW,
    });
    expect(result.storedToR2).toBe(false);
    expect(result.warnings.some((line) => line.includes('no R2 bucket'))).toBe(
      true
    );
    expect(db.rows.get('aaaaaaaaaaaaaaaa')?.r2_key).toBe(null);
  });

  test('missing D1 skips the row with a warning', async () => {
    const pair = makePair({ artifactId: 'aaaaaaaaaaaaaaaa' });
    const r2 = new FakeR2();
    const result = await storeBettiWasmArtifact({
      r2,
      wasmBytes: pair.wasmBytes,
      manifestJson: pair.manifestJson,
      now: FIXED_NOW,
    });
    expect(result.storedToD1).toBe(false);
    expect(result.storedToR2).toBe(true);
    expect(
      result.warnings.some((line) => line.includes('no D1 database'))
    ).toBe(true);
  });

  test('tampered bytes never reach D1 or R2', async () => {
    const pair = makePair({ artifactId: 'aaaaaaaaaaaaaaaa' });
    const tampered = Uint8Array.from(pair.wasmBytes);
    tampered[20] ^= 0x01;
    const db = new FakeD1();
    const r2 = new FakeR2();
    let failure: unknown = null;
    try {
      await storeBettiWasmArtifact({
        db,
        r2,
        wasmBytes: tampered,
        manifestJson: pair.manifestJson,
      });
    } catch (error) {
      failure = error;
    }
    expect(failure).not.toBe(null);
    expect(db.rows.size).toBe(0);
    expect(r2.objects.size).toBe(0);
  });

  test('loads by artifactId with verified bytes, and by key (newest wins)', async () => {
    const first = makePair({ artifactId: 'aaaaaaaaaaaaaaaa' });
    const second = makePair({ artifactId: 'bbbbbbbbbbbbbbbb' });
    const db = new FakeD1();
    const r2 = new FakeR2();

    await storeBettiWasmArtifact({
      db,
      r2,
      wasmBytes: first.wasmBytes,
      manifestJson: first.manifestJson,
      now: () => new Date('2026-07-01T00:00:00.000Z'),
    });
    await storeBettiWasmArtifact({
      db,
      r2,
      wasmBytes: second.wasmBytes,
      manifestJson: second.manifestJson,
      now: () => new Date('2026-07-02T00:00:00.000Z'),
    });

    const byId = await loadBettiWasmArtifact({
      db,
      r2,
      artifactId: 'aaaaaaaaaaaaaaaa',
    });
    expect(byId?.manifest.artifactId).toBe('aaaaaaaaaaaaaaaa');
    expect(byId?.wasmBytes?.length).toBe(first.wasmBytes.length);

    // Both pairs share filePath/functionName → same key; newest row wins.
    const byKey = await loadBettiWasmArtifact({
      db,
      key: fragmentKey(first.manifest.filePath, first.manifest.functionName),
    });
    expect(byKey?.row.artifact_id).toBe('bbbbbbbbbbbbbbbb');
    expect(byKey?.wasmBytes).toBe(undefined);

    const missing = await loadBettiWasmArtifact({
      db,
      artifactId: 'cccccccccccccccc',
    });
    expect(missing).toBe(null);
  });

  test('a corrupt R2 object fails closed on load', async () => {
    const pair = makePair({ artifactId: 'aaaaaaaaaaaaaaaa' });
    const db = new FakeD1();
    const r2 = new FakeR2();
    await storeBettiWasmArtifact({
      db,
      r2,
      wasmBytes: pair.wasmBytes,
      manifestJson: pair.manifestJson,
    });
    const key = bettiWasmModuleKey('aaaaaaaaaaaaaaaa');
    const corrupted = Uint8Array.from(r2.objects.get(key)!);
    corrupted[10] ^= 0xff;
    r2.objects.set(key, corrupted);

    let failure: unknown = null;
    try {
      await loadBettiWasmArtifact({
        db,
        r2,
        artifactId: 'aaaaaaaaaaaaaaaa',
      });
    } catch (error) {
      failure = error;
    }
    expect(String(failure)).toContain('integrity failure');
  });

  test('load requires exactly one selector', async () => {
    const db = new FakeD1();
    for (const selectors of [
      {},
      { artifactId: 'aaaaaaaaaaaaaaaa', key: 'deadbeefdeadbeef' },
    ]) {
      let failure: unknown = null;
      try {
        await loadBettiWasmArtifact({ db, ...selectors });
      } catch (error) {
        failure = error;
      }
      expect(String(failure)).toContain('exactly one selector');
    }
  });
});

describe('betti wasm store — Wave D3 owner/price migration', () => {
  test('a pre-D3 table gains owner_did/price_cents via guarded ADD COLUMN, idempotently', async () => {
    const db = new FakeD1(PRE_D3_COLUMNS);
    expect(db.columns.has('owner_did')).toBe(false);
    expect(db.columns.has('price_cents')).toBe(false);

    await ensureBettiWasmArtifactsTable(db);
    expect(db.columns.has('owner_did')).toBe(true);
    expect(db.columns.has('price_cents')).toBe(true);

    // Re-ensure: duplicate-column failures are swallowed, nothing else changes.
    await ensureBettiWasmArtifactsTable(db);
    expect(db.columns.size).toBe(PRE_D3_COLUMNS.length + 2);
  });

  test('new rows round-trip owner_did/price_cents on an upgraded table', async () => {
    const pair = makePair({ artifactId: 'aaaaaaaaaaaaaaaa' });
    const db = new FakeD1(PRE_D3_COLUMNS); // old table — store must migrate it
    const r2 = new FakeR2();

    const result = await storeBettiWasmArtifact({
      db,
      r2,
      wasmBytes: pair.wasmBytes,
      manifestJson: pair.manifestJson,
      ownerDid: 'did:key:zOwner123',
      priceCents: 5,
      now: FIXED_NOW,
    });
    expect(result.row.owner_did).toBe('did:key:zOwner123');
    expect(result.row.price_cents).toBe(5);

    const loaded = await loadBettiWasmArtifact({
      db,
      artifactId: 'aaaaaaaaaaaaaaaa',
    });
    expect(loaded?.row.owner_did).toBe('did:key:zOwner123');
    expect(loaded?.row.price_cents).toBe(5);
  });

  test('rows stored without owner/price keep NULLs (backward compat)', async () => {
    const pair = makePair({ artifactId: 'aaaaaaaaaaaaaaaa' });
    const db = new FakeD1();
    await storeBettiWasmArtifact({
      db,
      wasmBytes: pair.wasmBytes,
      manifestJson: pair.manifestJson,
      now: FIXED_NOW,
    });
    expect(db.rows.get('aaaaaaaaaaaaaaaa')?.owner_did).toBe(null);
    expect(db.rows.get('aaaaaaaaaaaaaaaa')?.price_cents).toBe(null);
  });

  test('pre-migration rows (no owner/price keys at all) load as explicit NULLs', async () => {
    const pair = makePair({ artifactId: 'aaaaaaaaaaaaaaaa' });
    const db = new FakeD1();
    await storeBettiWasmArtifact({
      db,
      wasmBytes: pair.wasmBytes,
      manifestJson: pair.manifestJson,
      now: FIXED_NOW,
    });
    // Simulate a row written before the migration: strip the new keys.
    const stored = db.rows.get('aaaaaaaaaaaaaaaa')!;
    const legacy = { ...stored } as Record<string, unknown>;
    delete legacy['owner_did'];
    delete legacy['price_cents'];
    db.rows.set('aaaaaaaaaaaaaaaa', legacy as unknown as BettiWasmArtifactRow);

    const loaded = await loadBettiWasmArtifact({
      db,
      artifactId: 'aaaaaaaaaaaaaaaa',
    });
    expect(loaded?.row.owner_did).toBe(null);
    expect(loaded?.row.price_cents).toBe(null);
  });

  test('file registry records ownerDid/priceCents only when supplied', async () => {
    const dir = await mkdtemp(join(tmpdir(), 'betti-wasm-registry-'));
    try {
      const owned = makePair({ artifactId: 'aaaaaaaaaaaaaaaa' });
      const unowned = makePair({ artifactId: 'bbbbbbbbbbbbbbbb' });

      const storedOwned = await storeBettiWasmArtifactToDirectory({
        dir,
        wasmBytes: owned.wasmBytes,
        manifestJson: owned.manifestJson,
        ownerDid: 'did:key:zOwner123',
        priceCents: 5,
        now: FIXED_NOW,
      });
      expect(storedOwned.entry.ownerDid).toBe('did:key:zOwner123');
      expect(storedOwned.entry.priceCents).toBe(5);

      const storedUnowned = await storeBettiWasmArtifactToDirectory({
        dir,
        wasmBytes: unowned.wasmBytes,
        manifestJson: unowned.manifestJson,
        now: FIXED_NOW,
      });
      expect('ownerDid' in storedUnowned.entry).toBe(false);
      expect('priceCents' in storedUnowned.entry).toBe(false);

      const registry = JSON.parse(
        await readFile(storedOwned.registryPath, 'utf8')
      ) as {
        artifacts: {
          artifactId: string;
          ownerDid?: string;
          priceCents?: number;
        }[];
      };
      const ownedEntry = registry.artifacts.find(
        (entry) => entry.artifactId === 'aaaaaaaaaaaaaaaa'
      );
      expect(ownedEntry?.ownerDid).toBe('did:key:zOwner123');
      expect(ownedEntry?.priceCents).toBe(5);
    } finally {
      await rm(dir, { recursive: true, force: true });
    }
  });
});

describe('betti wasm store — file registry mode', () => {
  test('roundtrips deterministically (sorted, idempotent, R2 key layout)', async () => {
    const dir = await mkdtemp(join(tmpdir(), 'betti-wasm-registry-'));
    try {
      const second = makePair({ artifactId: 'bbbbbbbbbbbbbbbb' });
      const first = makePair({ artifactId: 'aaaaaaaaaaaaaaaa' });

      // Store out of order — the index must still sort by artifactId.
      await storeBettiWasmArtifactToDirectory({
        dir,
        wasmBytes: second.wasmBytes,
        manifestJson: second.manifestJson,
        now: FIXED_NOW,
      });
      const stored = await storeBettiWasmArtifactToDirectory({
        dir,
        wasmBytes: first.wasmBytes,
        manifestJson: first.manifestJson,
        now: FIXED_NOW,
      });

      expect(stored.entry.wasmPath).toBe('betti/v1/aaaaaaaaaaaaaaaa.wasm');
      const storedBytes = new Uint8Array(await readFile(stored.wasmPath));
      expect(storedBytes.length).toBe(first.wasmBytes.length);
      expect(await readFile(stored.manifestPath, 'utf8')).toBe(
        first.manifestJson
      );

      const registryText = await readFile(stored.registryPath, 'utf8');
      const registry = JSON.parse(registryText) as {
        version: number;
        artifacts: { artifactId: string }[];
      };
      expect(registry.version).toBe(1);
      expect(registry.artifacts.map((entry) => entry.artifactId)).toEqual([
        'aaaaaaaaaaaaaaaa',
        'bbbbbbbbbbbbbbbb',
      ]);

      // Idempotent re-store with the same clock → identical index bytes.
      await storeBettiWasmArtifactToDirectory({
        dir,
        wasmBytes: first.wasmBytes,
        manifestJson: first.manifestJson,
        now: FIXED_NOW,
      });
      expect(await readFile(stored.registryPath, 'utf8')).toBe(registryText);
    } finally {
      await rm(dir, { recursive: true, force: true });
    }
  });

  test('file registry rejects tampered pairs before writing', async () => {
    const dir = await mkdtemp(join(tmpdir(), 'betti-wasm-registry-'));
    try {
      const pair = makePair({ artifactId: 'aaaaaaaaaaaaaaaa' });
      const tampered = Uint8Array.from(pair.wasmBytes);
      tampered[12] ^= 0x10;
      let failure: unknown = null;
      try {
        await storeBettiWasmArtifactToDirectory({
          dir,
          wasmBytes: tampered,
          manifestJson: pair.manifestJson,
        });
      } catch (error) {
        failure = error;
      }
      expect(String(failure)).toContain('integrity failure');
      let registryFailure: unknown = null;
      try {
        await readFile(join(dir, 'registry.json'), 'utf8');
      } catch (error) {
        registryFailure = error;
      }
      expect(registryFailure).not.toBe(null);
    } finally {
      await rm(dir, { recursive: true, force: true });
    }
  });
});
