// @a0n/gnosis/src/__tests__/betti-wasm-manifest.test.ts
import { describe, expect, test } from '@a0n/gnosis/test';
import {
  BETTI_WASM_MANIFEST_CUSTOM_SECTION,
  parseBettiManifestCustomSection,
  toRegisterWasmArtifactRequest,
  validateBettiWasmManifest,
  type BettiWasmManifest,
} from '../betti-wasm-manifest';
import type { RegisterWasmArtifactRequest } from '../wasm-execution';

const VALID_MANIFEST: BettiWasmManifest = {
  artifact_version: 'betti-wasm/1',
  artifactId: '0123456789abcdef',
  topologyHash: 'fedcba9876543210',
  sourceHash: 'b'.repeat(64),
  language: 'typescript',
  filePath: 'gnode/benchmarks/echo.ts',
  functionName: 'echo',
  certified_class: 'echo-fixture',
  tier: 'certified-direct',
  export_name: 'main',
  abi: 'json-v1',
  memory_export: 'memory',
  alloc_export: 'gnosis_alloc',
  result_length_export: 'gnosis_result_len',
  free_export: 'gnosis_free',
  imports: [],
  pure: true,
  limits: { memoryPagesMax: 64, fuel: 2 ** 42 },
  byteLength: 128,
  sha256: 'a'.repeat(64),
  decline_semantics: 'trap',
  wasm_opt: false,
};

function encodeLeb128U32(value: number): number[] {
  const bytes: number[] = [];
  let remaining = value >>> 0;
  do {
    let byte = remaining & 0x7f;
    remaining >>>= 7;
    if (remaining !== 0) {
      byte |= 0x80;
    }
    bytes.push(byte);
  } while (remaining !== 0);
  return bytes;
}

function buildCustomSection(name: string, payload: Uint8Array): number[] {
  const nameBytes = Array.from(new TextEncoder().encode(name));
  const body = [
    ...encodeLeb128U32(nameBytes.length),
    ...nameBytes,
    ...payload,
  ];
  return [0x00, ...encodeLeb128U32(body.length), ...body];
}

function buildWasmModule(sections: readonly number[][]): Uint8Array {
  return Uint8Array.from([
    0x00, 0x61, 0x73, 0x6d, // magic "\0asm"
    0x01, 0x00, 0x00, 0x00, // version 1
    ...sections.flat(),
  ]);
}

function manifestModule(manifest: unknown): Uint8Array {
  const payload = new TextEncoder().encode(JSON.stringify(manifest));
  return buildWasmModule([
    buildCustomSection(BETTI_WASM_MANIFEST_CUSTOM_SECTION, payload),
  ]);
}

describe('validateBettiWasmManifest', () => {
  test('accepts a well-formed json-v1 certified-direct manifest', () => {
    const manifest = validateBettiWasmManifest({ ...VALID_MANIFEST });
    expect(manifest).toEqual(VALID_MANIFEST);
  });

  test('accepts interpreter-embed manifests with declared imports', () => {
    const manifest = validateBettiWasmManifest({
      ...VALID_MANIFEST,
      tier: 'interpreter-embed',
      imports: [{ module: 'env', name: 'gnosis_host', kind: 'function' }],
    });
    expect(manifest.tier).toBe('interpreter-embed');
    expect(manifest.imports).toHaveLength(1);
  });

  test('rejects a wrong artifact_version', () => {
    expect(() =>
      validateBettiWasmManifest({ ...VALID_MANIFEST, artifact_version: 'betti-wasm/2' })
    ).toThrow(/artifact_version must be 'betti-wasm\/1'/);
  });

  test('rejects non-empty imports on certified-direct', () => {
    expect(() =>
      validateBettiWasmManifest({
        ...VALID_MANIFEST,
        imports: [{ module: 'env', name: 'gnosis_host', kind: 'function' }],
      })
    ).toThrow(/certified-direct' requires an empty import section/);
  });

  test('rejects malformed artifactId hex', () => {
    expect(() =>
      validateBettiWasmManifest({ ...VALID_MANIFEST, artifactId: '0123456789ABCDEF' })
    ).toThrow(/artifactId must be 16 lowercase hex/);
    expect(() =>
      validateBettiWasmManifest({ ...VALID_MANIFEST, artifactId: '0123' })
    ).toThrow(/artifactId must be 16 lowercase hex/);
  });

  test('rejects malformed sha256 hex', () => {
    expect(() =>
      validateBettiWasmManifest({ ...VALID_MANIFEST, sha256: 'Z'.repeat(64) })
    ).toThrow(/sha256 must be 64 lowercase hex/);
    expect(() =>
      validateBettiWasmManifest({ ...VALID_MANIFEST, sha256: 'a'.repeat(63) })
    ).toThrow(/sha256 must be 64 lowercase hex/);
  });

  test('rejects a missing abi', () => {
    const { abi: _abi, ...withoutAbi } = VALID_MANIFEST;
    expect(() => validateBettiWasmManifest(withoutAbi)).toThrow(
      /abi must be one of 'scalar-v1' \| 'json-v1'/
    );
  });

  test('rejects decline semantics other than trap', () => {
    expect(() =>
      validateBettiWasmManifest({ ...VALID_MANIFEST, decline_semantics: 'exit' })
    ).toThrow(/decline_semantics must be 'trap'/);
  });

  test('rejects non-object input', () => {
    expect(() => validateBettiWasmManifest(null)).toThrow(/expected an object/);
    expect(() => validateBettiWasmManifest('manifest')).toThrow(/expected an object/);
  });
});

describe('parseBettiManifestCustomSection', () => {
  test('round-trips a manifest embedded in a minimal wasm module', () => {
    const wasmBytes = manifestModule(VALID_MANIFEST);
    const parsed = parseBettiManifestCustomSection(wasmBytes);
    expect(parsed).toEqual(VALID_MANIFEST);
  });

  test('skips unrelated custom sections before the manifest', () => {
    const payload = new TextEncoder().encode(JSON.stringify(VALID_MANIFEST));
    const wasmBytes = buildWasmModule([
      buildCustomSection('name', new TextEncoder().encode('decoy')),
      buildCustomSection(BETTI_WASM_MANIFEST_CUSTOM_SECTION, payload),
    ]);
    expect(parseBettiManifestCustomSection(wasmBytes)).toEqual(VALID_MANIFEST);
  });

  test('returns null when no gnosis-betti-manifest section exists', () => {
    const wasmBytes = buildWasmModule([
      buildCustomSection('name', new TextEncoder().encode('decoy')),
    ]);
    expect(parseBettiManifestCustomSection(wasmBytes)).toBeNull();
  });

  test('returns null for a bare header module', () => {
    expect(parseBettiManifestCustomSection(buildWasmModule([]))).toBeNull();
  });

  test('fails closed on bad magic', () => {
    const wasmBytes = manifestModule(VALID_MANIFEST);
    wasmBytes[0] = 0x7f;
    expect(() => parseBettiManifestCustomSection(wasmBytes)).toThrow(/bad magic/);
  });

  test('fails closed on a truncated section', () => {
    const wasmBytes = manifestModule(VALID_MANIFEST);
    expect(() =>
      parseBettiManifestCustomSection(wasmBytes.subarray(0, wasmBytes.length - 4))
    ).toThrow(/declares .* byte/);
  });

  test('fails closed on an invalid embedded manifest', () => {
    const wasmBytes = manifestModule({
      ...VALID_MANIFEST,
      artifact_version: 'betti-wasm/2',
    });
    expect(() => parseBettiManifestCustomSection(wasmBytes)).toThrow(
      /artifact_version must be 'betti-wasm\/1'/
    );
  });
});

describe('toRegisterWasmArtifactRequest', () => {
  test('produces a RegisterWasmArtifactRequest with explicit artifactId', () => {
    // Type-level check: the helper's output must satisfy the exact request
    // type accepted by GnosisWasmExecutionSurface.register.
    const request: RegisterWasmArtifactRequest = toRegisterWasmArtifactRequest(
      VALID_MANIFEST,
      { wasmBase64: 'AGFzbQEAAAA=' }
    );

    expect(request).toEqual({
      artifactId: '0123456789abcdef',
      wasmBase64: 'AGFzbQEAAAA=',
      exportName: 'main',
      abi: 'json-v1',
      memoryExport: 'memory',
      allocExport: 'gnosis_alloc',
      resultLengthExport: 'gnosis_result_len',
      freeExport: 'gnosis_free',
    });
  });

  test('carries a wasmPath when bytes live on disk', () => {
    const request: RegisterWasmArtifactRequest = toRegisterWasmArtifactRequest(
      VALID_MANIFEST,
      { wasmPath: '/betti/v1/0123456789abcdef.wasm' }
    );
    expect(request.wasmPath).toBe('/betti/v1/0123456789abcdef.wasm');
    expect(request.wasmBase64).toBeUndefined();
    expect(request.artifactId).toBe(VALID_MANIFEST.artifactId);
  });

  test('omits optional exports absent from the manifest', () => {
    const scalarManifest = validateBettiWasmManifest({
      ...VALID_MANIFEST,
      abi: 'scalar-v1',
      memory_export: undefined,
      alloc_export: undefined,
      result_length_export: undefined,
      free_export: undefined,
    });
    const request = toRegisterWasmArtifactRequest(scalarManifest, {
      wasmBase64: 'AGFzbQEAAAA=',
    });
    expect(request.abi).toBe('scalar-v1');
    expect('memoryExport' in request).toBe(false);
    expect('allocExport' in request).toBe(false);
    expect('resultLengthExport' in request).toBe(false);
    expect('freeExport' in request).toBe(false);
  });

  test('fails closed without artifact bytes', () => {
    expect(() => toRegisterWasmArtifactRequest(VALID_MANIFEST, {})).toThrow(
      /requires artifact bytes/
    );
  });

  test('fails closed on ambiguous artifact bytes', () => {
    expect(() =>
      toRegisterWasmArtifactRequest(VALID_MANIFEST, {
        wasmBase64: 'AGFzbQEAAAA=',
        wasmPath: '/betti/v1/0123456789abcdef.wasm',
      })
    ).toThrow(/not both/);
  });
});
