// @a0n/gnosis/src/__tests__/betti-wasm-execution.test.ts
/**
 * Betti → WASM end-to-end contract test — the "no new glue" proof.
 *
 * When the Betti emitter binary (monster/gnexec) is resolvable, this emits
 * the echo fixture with `--emit-wasm-out`, validates the `betti-wasm/1`
 * manifest sidecar, and runs the artifact through the existing
 * `GnosisWasmExecutionSurface.register` + `execute` path via
 * `toRegisterWasmArtifactRequest` — zero artifact-specific glue. When the
 * binary is absent the emitter-backed tests skip politely (log + pass).
 *
 * The pre-compiled `WebAssembly.Module` tests always run: compiling bytes
 * in-test via `new WebAssembly.Module` is allowed under Node (only the
 * Cloudflare edge forbids runtime compile).
 */
import { execFile } from 'node:child_process';
import { existsSync } from 'node:fs';
import { mkdtemp, readFile, rm } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { delimiter, dirname, join, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import { promisify } from 'node:util';
import { describe, expect, test } from '@a0n/gnosis/test';
import {
  parseBettiManifestCustomSection,
  toRegisterWasmArtifactRequest,
  validateBettiWasmManifest,
  type BettiWasmManifest,
} from '../betti-wasm-manifest';
import { GnosisWasmExecutionSurface } from '../wasm-execution';

const execFileAsync = promisify(execFile);

const HERE = dirname(fileURLToPath(import.meta.url));
const POLYGLOT_TARGET_DIR = resolve(HERE, '../../polyglot/target');
const ECHO_FIXTURE = resolve(HERE, '../../gnode/benchmarks/echo.ts');

/**
 * Same resolution order as x-gnosis `MonsterEmitWasmProvider`:
 * $GNOSIS_MONSTER_BIN → release/monster → release/gnexec → debug/gnexec →
 * `monster` on PATH.
 */
function resolveEmitterBinary(): string | null {
  const fromEnv = process.env.GNOSIS_MONSTER_BIN;
  if (fromEnv && fromEnv.length > 0 && existsSync(fromEnv)) {
    return fromEnv;
  }

  for (const candidate of [
    join(POLYGLOT_TARGET_DIR, 'release', 'monster'),
    join(POLYGLOT_TARGET_DIR, 'release', 'gnexec'),
    join(POLYGLOT_TARGET_DIR, 'debug', 'gnexec'),
  ]) {
    if (existsSync(candidate)) {
      return candidate;
    }
  }

  const pathEnv = process.env.PATH ?? '';
  for (const dir of pathEnv.split(delimiter)) {
    if (dir.length > 0 && existsSync(join(dir, 'monster'))) {
      return join(dir, 'monster');
    }
  }

  return null;
}

const EMITTER_BINARY = resolveEmitterBinary();

function logEmitterSkip(testName: string): void {
  console.log(
    `betti-wasm-execution: skipping '${testName}' — no monster/gnexec emitter binary ` +
      '(build with `cargo build --bin gnexec` in open-source/gnosis/polyglot).'
  );
}

interface EmittedEchoArtifact {
  readonly wasmBytes: Uint8Array;
  readonly manifest: BettiWasmManifest;
}

async function emitEchoArtifact(
  binary: string,
  workDir: string
): Promise<EmittedEchoArtifact> {
  const modulePath = join(workDir, 'module.wasm');
  await execFileAsync(binary, [
    ECHO_FIXTURE,
    '--export',
    'app',
    '--emit-wasm-out',
    modulePath,
  ]);

  const wasmBytes = new Uint8Array(await readFile(modulePath));
  const manifest = validateBettiWasmManifest(
    JSON.parse(await readFile(join(workDir, 'module.manifest.json'), 'utf8'))
  );
  return { wasmBytes, manifest };
}

// Hand-assembled scalar-v1 module: (func (export "main") (param f64)
// (result f64) local.get 0 f64.const 2 f64.mul) — doubles its argument.
const SCALAR_DOUBLE_WASM = Uint8Array.from([
  0x00, 0x61, 0x73, 0x6d, 0x01, 0x00, 0x00, 0x00, // header
  0x01, 0x06, 0x01, 0x60, 0x01, 0x7c, 0x01, 0x7c, // type: (f64) -> f64
  0x03, 0x02, 0x01, 0x00, // function section
  0x07, 0x08, 0x01, 0x04, 0x6d, 0x61, 0x69, 0x6e, 0x00, 0x00, // export "main"
  0x0a, 0x10, 0x01, 0x0e, 0x00, 0x20, 0x00, 0x44, // code: local.get 0,
  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x40, // f64.const 2,
  0xa2, 0x0b, // f64.mul, end
]);

describe('betti wasm end-to-end contract', () => {
  test('emitted echo artifact executes through the existing wasm surface', async () => {
    if (!EMITTER_BINARY) {
      logEmitterSkip('emitted echo artifact executes');
      return;
    }

    const workDir = await mkdtemp(join(tmpdir(), 'betti-wasm-contract-'));
    try {
      const { wasmBytes, manifest } = await emitEchoArtifact(
        EMITTER_BINARY,
        workDir
      );

      // Self-describing artifact: the embedded custom section carries the
      // same manifest as the sidecar.
      expect(parseBettiManifestCustomSection(wasmBytes)).toEqual(manifest);

      const surface = new GnosisWasmExecutionSurface();
      const artifact = surface.register(
        toRegisterWasmArtifactRequest(manifest, {
          wasmBase64: Buffer.from(wasmBytes).toString('base64'),
        })
      );
      expect(artifact.artifactId).toBe(manifest.artifactId);

      const response = await surface.execute({
        artifactId: manifest.artifactId,
        input: { name: 'gnode' },
      });
      expect(response.payload).toBe('hello:gnode');
      expect(response.abi).toBe(manifest.abi);
      expect(response.exportName).toBe(manifest.export_name);
    } finally {
      await rm(workDir, { recursive: true, force: true });
    }
  });

  test('decline path (trap) surfaces as a failed execution', async () => {
    if (!EMITTER_BINARY) {
      logEmitterSkip('decline path traps');
      return;
    }

    const workDir = await mkdtemp(join(tmpdir(), 'betti-wasm-decline-'));
    try {
      const { wasmBytes, manifest } = await emitEchoArtifact(
        EMITTER_BINARY,
        workDir
      );
      expect(manifest.decline_semantics).toBe('trap');

      const surface = new GnosisWasmExecutionSurface();
      surface.register(
        toRegisterWasmArtifactRequest(manifest, {
          wasmBase64: Buffer.from(wasmBytes).toString('base64'),
        })
      );

      let failure: unknown = null;
      try {
        await surface.execute({
          artifactId: manifest.artifactId,
          input: { nope: 1 },
        });
      } catch (error) {
        failure = error;
      }
      expect(failure).not.toBe(null);
    } finally {
      await rm(workDir, { recursive: true, force: true });
    }
  });

  test('a pre-compiled WebAssembly.Module registers and executes', async () => {
    const module = new WebAssembly.Module(SCALAR_DOUBLE_WASM);
    const surface = new GnosisWasmExecutionSurface();
    const artifact = surface.register({
      artifactId: 'precompiled-scalar-double',
      wasmModule: module,
      exportName: 'main',
      abi: 'scalar-v1',
    });
    expect(artifact.artifactId).toBe('precompiled-scalar-double');

    const response = await surface.execute({
      artifactId: 'precompiled-scalar-double',
      input: 21,
    });
    expect(response.payload).toBe(42);
    expect(surface.getStatus().artifacts[0]?.source).toBe('module');
  });

  test('a moduleResolver-backed artifact registers without bytes and executes', async () => {
    const module = new WebAssembly.Module(SCALAR_DOUBLE_WASM);
    const surface = new GnosisWasmExecutionSurface({
      moduleResolver: (descriptor) =>
        descriptor.artifactId === 'resolver-scalar-double' ? module : null,
    });
    const artifact = surface.register({
      artifactId: 'resolver-scalar-double',
      exportName: 'main',
      abi: 'scalar-v1',
    });
    expect(artifact.artifactId).toBe('resolver-scalar-double');

    const response = await surface.execute({
      artifactId: 'resolver-scalar-double',
      input: 4,
    });
    expect(response.payload).toBe(8);
    expect(surface.getStatus().artifacts[0]?.source).toBe('resolver');
  });

  test('registering a wasmModule without an explicit artifactId fails closed', () => {
    const module = new WebAssembly.Module(SCALAR_DOUBLE_WASM);
    const surface = new GnosisWasmExecutionSurface();
    expect(() =>
      surface.register({
        wasmModule: module,
        exportName: 'main',
        abi: 'scalar-v1',
      })
    ).toThrow(/explicit artifactId/);
  });
});
