// @a0n/gnosis/src/__tests__/polyglot-integration.test.ts
import { describe, test, expect } from '@a0n/gnosis/test';
import { execFileSync } from 'node:child_process';
import * as path from 'node:path';
import * as fs from 'node:fs';
import {
  invokeSubprocess,
  invokeFunction,
  type InvokeRequest,
} from '../runtime/subprocess-invoker';
import {
  registerPolyglotBridgeHandlers,
  registerMultiLanguageBridgeHandlers,
  buildMultiLanguageManifest,
  type PolyglotExecutionManifest,
} from '../polyglot-execution-bridge';
import { GnosisRegistry } from '../runtime/registry';
import { buildExecutionTrace, formatExecutionTrace } from '../polyglot-trace';
import {
  compose,
  translate,
  findBestLanguage,
  computeBetaCost,
  extractFunctions,
  generateTopoRaceTopology,
} from '../polyglot-compose';
import { scaffoldTopology } from '../polyglot-scaffold';
import {
  PolyglotStrategyMemory,
  type LanguageObservation,
} from '../polyglot-memory';
import {
  analyzeForOptimizations,
  TopologyProfiler,
} from '../polyglot-optimizer';
import type { ASTNode, ASTEdge, GraphAST } from '../betty/compiler';

const FIXTURES_DIR = path.resolve(__dirname, '../../harnesses/fixtures');

interface ScriptServiceFixture {
  readonly language: string;
  readonly fileName: string;
  readonly sourceText: string;
  readonly expectedBody: string;
  readonly isAvailable: () => boolean;
  readonly timeoutMs?: number;
}

function isCommandAvailable(command: string): boolean {
  try {
    execFileSync('sh', ['-lc', `command -v ${command}`], {
      stdio: 'ignore',
    });
    return true;
  } catch {
    return false;
  }
}

function decodeCompatibilityBody(responseValue: unknown): string | null {
  const body = (responseValue as { response?: { body?: { value?: string } } })
    ?.response?.body;

  if (!body?.value) {
    return null;
  }

  return Buffer.from(body.value, 'base64').toString('utf8');
}

function removeIfPresent(filePath: string): void {
  try {
    fs.unlinkSync(filePath);
  } catch (error) {
    const errno = error as NodeJS.ErrnoException;
    if (errno.code !== 'ENOENT') {
      throw error;
    }
  }
}

async function withSubprocessTimeout<T>(
  timeoutMs: number,
  fn: () => Promise<T>
): Promise<T> {
  const previousTimeout = process.env.GNODE_SUBPROCESS_TIMEOUT_MS;
  process.env.GNODE_SUBPROCESS_TIMEOUT_MS = String(timeoutMs);

  try {
    return await fn();
  } finally {
    if (previousTimeout === undefined) {
      delete process.env.GNODE_SUBPROCESS_TIMEOUT_MS;
    } else {
      process.env.GNODE_SUBPROCESS_TIMEOUT_MS = previousTimeout;
    }
  }
}

function normalizeFixtureSource(sourceText: string): string {
  const trimmed = sourceText.replace(/^\s*\n/, '').replace(/\s+$/, '');
  const lines = trimmed.split('\n');
  const indentation = lines
    .filter((line) => line.trim().length > 0)
    .reduce((minimumIndentation, line) => {
      const nextIndentation = line.match(/^\s*/)?.[0].length ?? 0;
      return Math.min(minimumIndentation, nextIndentation);
    }, Number.POSITIVE_INFINITY);

  const normalizedLines = lines.map((line) =>
    Number.isFinite(indentation) ? line.slice(indentation) : line
  );

  return `${normalizedLines.join('\n')}\n`;
}

let scriptServiceFixtureRunId = 0;

function makeScriptServiceTempFilePath(fixture: ScriptServiceFixture): string {
  scriptServiceFixtureRunId += 1;

  const safeLanguage = fixture.language.replace(/[^a-z0-9_-]+/gi, '-');
  const uniqueSuffix = `${process.pid}-${Date.now()}-${scriptServiceFixtureRunId}`;

  return path.join(
    FIXTURES_DIR,
    `script-service-${safeLanguage}-${uniqueSuffix}-${fixture.fileName}`
  );
}

const SCRIPT_SERVICE_FIXTURES: readonly ScriptServiceFixture[] = [
  {
    language: 'rust',
    fileName: 'standard_service.rs',
    sourceText: `
      use std::env;

      fn main() {
        print!(
          "rust:{}",
          env::var("XGNOSIS_HTTP_PATH").unwrap_or_default()
        );
      }
    `,
    expectedBody: 'rust:/standard',
    isAvailable: () => isCommandAvailable('rustc'),
  },
  {
    language: 'c',
    fileName: 'standard_service.c',
    sourceText: `
      #include <stdio.h>
      #include <stdlib.h>

      int main(void) {
        const char *path = getenv("XGNOSIS_HTTP_PATH");
        printf("c:%s", path ? path : "");
        return 0;
      }
    `,
    expectedBody: 'c:/standard',
    isAvailable: () => isCommandAvailable('cc'),
  },
  {
    language: 'cpp',
    fileName: 'standard_service.cpp',
    sourceText: `
      #include <cstdlib>
      #include <iostream>

      int main() {
        const char *path = std::getenv("XGNOSIS_HTTP_PATH");
        std::cout << "cpp:" << (path ? path : "");
        return 0;
      }
    `,
    expectedBody: 'cpp:/standard',
    isAvailable: () => isCommandAvailable('c++'),
  },
  {
    language: 'c_sharp',
    fileName: 'Program.cs',
    sourceText: `
      using System;

      Console.Write(
        "c_sharp:" + (Environment.GetEnvironmentVariable("XGNOSIS_HTTP_PATH") ?? "")
      );
    `,
    expectedBody: 'c_sharp:/standard',
    isAvailable: () =>
      isCommandAvailable('dotnet') || isCommandAvailable('csc'),
    timeoutMs: 120000,
  },
  {
    language: 'php',
    fileName: 'standard_service.php',
    sourceText: `
      <?php
      fwrite(STDOUT, 'php:' . getenv('XGNOSIS_HTTP_PATH'));
    `,
    expectedBody: 'php:/standard',
    isAvailable: () => isCommandAvailable('php'),
  },
  {
    language: 'bash',
    fileName: 'standard_service.sh',
    sourceText: `
      #!/usr/bin/env bash
      printf 'bash:%s' "$XGNOSIS_HTTP_PATH"
    `,
    expectedBody: 'bash:/standard',
    isAvailable: () => isCommandAvailable('bash'),
  },
  {
    language: 'scala',
    fileName: 'standard_service.scala',
    sourceText: `
      object StandardService extends App {
        print("scala:" + sys.env.getOrElse("XGNOSIS_HTTP_PATH", ""))
      }
    `,
    expectedBody: 'scala:/standard',
    isAvailable: () =>
      isCommandAvailable('scala-cli') || isCommandAvailable('scala'),
    timeoutMs: 120000,
  },
  {
    language: 'kotlin',
    fileName: 'standard_service.kts',
    sourceText: `
      print("kotlin:" + (System.getenv("XGNOSIS_HTTP_PATH") ?: ""))
    `,
    expectedBody: 'kotlin:/standard',
    isAvailable: () => isCommandAvailable('kotlin'),
    timeoutMs: 120000,
  },
  {
    language: 'swift',
    fileName: 'standard_service.swift',
    sourceText: `
      import Foundation

      let path = ProcessInfo.processInfo.environment["XGNOSIS_HTTP_PATH"] ?? ""
      print("swift:\\(path)", terminator: "")
    `,
    expectedBody: 'swift:/standard',
    isAvailable: () => isCommandAvailable('swift'),
  },
  {
    language: 'haskell',
    fileName: 'standard_service.hs',
    sourceText: `
      import System.Environment (lookupEnv)

      main :: IO ()
      main = do
        path <- lookupEnv "XGNOSIS_HTTP_PATH"
        putStr ("haskell:" ++ maybe "" id path)
    `,
    expectedBody: 'haskell:/standard',
    isAvailable: () =>
      isCommandAvailable('runghc') || isCommandAvailable('ghc'),
    timeoutMs: 120000,
  },
  {
    language: 'ocaml',
    fileName: 'standard_service.ml',
    sourceText: `
      let path =
        match Sys.getenv_opt "XGNOSIS_HTTP_PATH" with
        | Some value -> value
        | None -> ""

      let () = print_string ("ocaml:" ^ path)
    `,
    expectedBody: 'ocaml:/standard',
    isAvailable: () => isCommandAvailable('ocaml'),
  },
  {
    language: 'lua',
    fileName: 'standard_service.lua',
    sourceText: `
      io.write("lua:" .. (os.getenv("XGNOSIS_HTTP_PATH") or ""))
    `,
    expectedBody: 'lua:/standard',
    isAvailable: () => isCommandAvailable('lua'),
  },
  {
    language: 'elixir',
    fileName: 'standard_service.exs',
    sourceText: `
      IO.write("elixir:" <> (System.get_env("XGNOSIS_HTTP_PATH") || ""))
    `,
    expectedBody: 'elixir:/standard',
    isAvailable: () => isCommandAvailable('elixir'),
  },
  {
    language: 'zig',
    fileName: 'standard_service.zig',
    sourceText: `
      const std = @import("std");

      pub fn main() !void {
          const stdout = std.fs.File.stdout().deprecatedWriter();
          if (std.posix.getenv("XGNOSIS_HTTP_PATH")) |path| {
              try stdout.print("zig:{s}", .{path});
          } else {
              try stdout.writeAll("zig:");
          }
      }
    `,
    expectedBody: 'zig:/standard',
    isAvailable: () => isCommandAvailable('zig'),
  },
];

// ─── Subprocess Harness Tests ────────────────────────────────────────────────

describe('subprocess invoker', () => {
  test('python harness executes function and returns result', async () => {
    const response = await invokeFunction(
      'python',
      path.join(FIXTURES_DIR, 'e2e_test.py'),
      'add',
      [3, 4]
    );
    expect(response.status).toBe('ok');
    expect(response.value).toBe(7);
  }, 10000);

  test('python harness handles complex return values', async () => {
    const response = await invokeFunction(
      'python',
      path.join(FIXTURES_DIR, 'e2e_test.py'),
      'main',
      []
    );
    expect(response.status).toBe('ok');
    expect(response.value).toMatchObject({
      add_result: 7,
      fib_10: 55,
    });
  }, 10000);

  test('python harness reports errors with mapped location', async () => {
    // Create a file with an intentional error.
    const tempFile = path.join(FIXTURES_DIR, '_error_test.py');
    fs.writeFileSync(tempFile, 'def broken(x):\n    return x / 0\n', 'utf8');

    try {
      const response = await invokeFunction('python', tempFile, 'broken', [42]);
      expect(response.status).toBe('error');
      expect(response.mappedError).toBeDefined();
      expect(response.mappedError!.language).toBe('python');
      expect(response.mappedError!.line).toBe(2); // division by zero on line 2
    } finally {
      removeIfPresent(tempFile);
    }
  }, 10000);

  test('node harness executes function', async () => {
    const tempFile = path.join(FIXTURES_DIR, '_node_test.mjs');
    fs.writeFileSync(
      tempFile,
      'export function multiply(a, b) { return a * b; }\n',
      'utf8'
    );

    try {
      const response = await invokeFunction(
        'javascript',
        tempFile,
        'multiply',
        [6, 7]
      );
      expect(response.status).toBe('ok');
      expect(response.value).toBe(42);
    } finally {
      removeIfPresent(tempFile);
    }
  }, 10000);

  test('node harness serves express compatibility requests with inline handlers', async () => {
    const tempFile = path.join(FIXTURES_DIR, '_express_compat_test.cjs');
    fs.writeFileSync(
      tempFile,
      `
        const express = require('express');
        const app = express();

        app.use((req, res, next) => {
          res.set('X-Compat', '1');
          next();
        });

        app.get('/users/:id', async (req, res) => {
          res.json({ id: req.params.id });
        });

        app.post('/echo', express.json(), async (req, res) => {
          res.status(201).json({ echoed: req.body });
        });

        app.listen(3000);
      `,
      'utf8'
    );

    const getRequest: InvokeRequest = {
      action: 'execute',
      language: 'javascript',
      filePath: tempFile,
      functionName: '__xgnosis_compat__',
      args: [],
      compatibility: {
        family: 'node-router',
        framework: 'express',
        request: {
          method: 'GET',
          path: '/users/42',
          headers: {
            host: 'compat.test',
          },
        },
      },
    };

    const postRequest: InvokeRequest = {
      action: 'execute',
      language: 'javascript',
      filePath: tempFile,
      functionName: '__xgnosis_compat__',
      args: [],
      compatibility: {
        family: 'node-router',
        framework: 'express',
        request: {
          method: 'POST',
          path: '/echo',
          headers: {
            host: 'compat.test',
            'content-type': 'application/json',
          },
          body: {
            encoding: 'base64',
            value: Buffer.from(JSON.stringify({ ok: true })).toString('base64'),
          },
        },
      },
    };

    try {
      const getResponse = await invokeSubprocess(getRequest);
      expect(getResponse.status).toBe('ok');
      expect(getResponse.value).toMatchObject({
        matched: true,
        response: {
          status: 200,
          headers: {
            'X-Compat': '1',
          },
        },
      });
      expect(
        JSON.parse(
          Buffer.from(
            (
              getResponse.value as {
                response: { body: { value: string } };
              }
            ).response.body.value,
            'base64'
          ).toString('utf8')
        )
      ).toEqual({
        id: '42',
      });

      const postResponse = await invokeSubprocess(postRequest);
      expect(postResponse.status).toBe('ok');
      expect(postResponse.value).toMatchObject({
        matched: true,
        response: {
          status: 201,
          headers: {
            'X-Compat': '1',
          },
        },
      });
      expect(
        JSON.parse(
          Buffer.from(
            (
              postResponse.value as {
                response: { body: { value: string } };
              }
            ).response.body.value,
            'base64'
          ).toString('utf8')
        )
      ).toEqual({
        echoed: { ok: true },
      });
    } finally {
      removeIfPresent(tempFile);
    }
  }, 10000);

  test('node harness serves Koa compatibility requests with router middleware', async () => {
    const tempFile = path.join(FIXTURES_DIR, '_koa_compat_test.cjs');
    fs.writeFileSync(
      tempFile,
      `
        const Koa = require('koa');
        const Router = require('@koa/router');

        const app = new Koa();
        const router = new Router();

        app.use(async (ctx, next) => {
          ctx.set('X-Koa-Compat', '1');
          await next();
        });

        router.get('/users/:id', async (ctx) => {
          ctx.body = {
            id: ctx.params.id,
            marker: ctx.query.marker,
          };
        });

        router.post('/users', async (ctx) => {
          ctx.status = 201;
          ctx.body = {
            echoed: ctx.request.body,
          };
        });

        app.use(router.routes());
        app.use(router.allowedMethods());
        app.listen(3000);
      `,
      'utf8'
    );

    const getRequest: InvokeRequest = {
      action: 'execute',
      language: 'javascript',
      filePath: tempFile,
      functionName: '__xgnosis_compat__',
      args: [],
      compatibility: {
        family: 'node-router',
        framework: 'koa',
        request: {
          method: 'GET',
          path: '/users/42',
          query: '?marker=1',
          headers: {
            host: 'compat.test',
          },
        },
      },
    };

    const postRequest: InvokeRequest = {
      action: 'execute',
      language: 'javascript',
      filePath: tempFile,
      functionName: '__xgnosis_compat__',
      args: [],
      compatibility: {
        family: 'node-router',
        framework: 'koa',
        request: {
          method: 'POST',
          path: '/users',
          headers: {
            host: 'compat.test',
            'content-type': 'application/json',
          },
          body: {
            encoding: 'base64',
            value: Buffer.from(JSON.stringify({ ok: true })).toString('base64'),
          },
        },
      },
    };

    try {
      const getResponse = await invokeSubprocess(getRequest);
      expect(getResponse.status).toBe('ok');
      expect(getResponse.value).toMatchObject({
        matched: true,
        response: {
          status: 200,
          headers: {
            'X-Koa-Compat': '1',
          },
        },
      });
      expect(
        JSON.parse(
          Buffer.from(
            (
              getResponse.value as {
                response: { body: { value: string } };
              }
            ).response.body.value,
            'base64'
          ).toString('utf8')
        )
      ).toEqual({
        id: '42',
        marker: '1',
      });

      const postResponse = await invokeSubprocess(postRequest);
      expect(postResponse.status).toBe('ok');
      expect(postResponse.value).toMatchObject({
        matched: true,
        response: {
          status: 201,
          headers: {
            'X-Koa-Compat': '1',
          },
        },
      });
      expect(
        JSON.parse(
          Buffer.from(
            (
              postResponse.value as {
                response: { body: { value: string } };
              }
            ).response.body.value,
            'base64'
          ).toString('utf8')
        )
      ).toEqual({
        echoed: { ok: true },
      });
    } finally {
      removeIfPresent(tempFile);
    }
  }, 10000);

  test('node harness serves Fastify compatibility requests with hooks and preHandlers', async () => {
    const tempFile = path.join(FIXTURES_DIR, '_fastify_compat_test.cjs');
    fs.writeFileSync(
      tempFile,
      `
        const fastify = require('fastify');
        const app = fastify();

        app.addHook('onRequest', async (request, reply) => {
          reply.header('X-Fastify-Compat', '1');
        });

        app.get('/users/:id', async (request, reply) => {
          return {
            id: request.params.id,
            marker: request.query.marker,
          };
        });

        app.post(
          '/users',
          {
            preHandler: async (request, reply) => {
              reply.header('X-Fastify-Pre', '1');
            },
          },
          async (request, reply) => {
            reply.code(201);
            return {
              echoed: request.body,
            };
          }
        );

        app.listen({ port: 3001 });
      `,
      'utf8'
    );

    const getRequest: InvokeRequest = {
      action: 'execute',
      language: 'javascript',
      filePath: tempFile,
      functionName: '__xgnosis_compat__',
      args: [],
      compatibility: {
        family: 'node-router',
        framework: 'fastify',
        request: {
          method: 'GET',
          path: '/users/42',
          query: '?marker=1',
          headers: {
            host: 'compat.test',
          },
        },
      },
    };

    const postRequest: InvokeRequest = {
      action: 'execute',
      language: 'javascript',
      filePath: tempFile,
      functionName: '__xgnosis_compat__',
      args: [],
      compatibility: {
        family: 'node-router',
        framework: 'fastify',
        request: {
          method: 'POST',
          path: '/users',
          headers: {
            host: 'compat.test',
            'content-type': 'application/json',
          },
          body: {
            encoding: 'base64',
            value: Buffer.from(JSON.stringify({ ok: true })).toString('base64'),
          },
        },
      },
    };

    try {
      const getResponse = await invokeSubprocess(getRequest);
      expect(getResponse.status).toBe('ok');
      expect(getResponse.value).toMatchObject({
        matched: true,
        response: {
          status: 200,
          headers: {
            'X-Fastify-Compat': '1',
          },
        },
      });
      expect(
        JSON.parse(
          Buffer.from(
            (
              getResponse.value as {
                response: { body: { value: string } };
              }
            ).response.body.value,
            'base64'
          ).toString('utf8')
        )
      ).toEqual({
        id: '42',
        marker: '1',
      });

      const postResponse = await invokeSubprocess(postRequest);
      expect(postResponse.status).toBe('ok');
      expect(postResponse.value).toMatchObject({
        matched: true,
        response: {
          status: 201,
          headers: {
            'X-Fastify-Compat': '1',
            'X-Fastify-Pre': '1',
          },
        },
      });
      expect(
        JSON.parse(
          Buffer.from(
            (
              postResponse.value as {
                response: { body: { value: string } };
              }
            ).response.body.value,
            'base64'
          ).toString('utf8')
        )
      ).toEqual({
        echoed: { ok: true },
      });
    } finally {
      removeIfPresent(tempFile);
    }
  }, 10000);

  test('go harness serves gin compatibility requests with middleware and JSON binding', async () => {
    const tempFile = path.join(FIXTURES_DIR, '_gin_compat_test.go');
    fs.writeFileSync(
      tempFile,
      `
        package main

        import (
          "net/http"
          "github.com/gin-gonic/gin"
        )

        type createUserInput struct {
          Name string \`json:"name"\`
        }

        func main() {
          r := gin.Default()

          r.Use(func(c *gin.Context) {
            c.Header("X-Gin-Compat", "1")
            c.Next()
          })

          r.GET("/users/:id", func(c *gin.Context) {
            c.JSON(http.StatusOK, gin.H{
              "id": c.Param("id"),
              "filter": c.Query("filter"),
            })
          })

          r.POST("/users", func(c *gin.Context) {
            var input createUserInput
            _ = c.ShouldBindJSON(&input)
            c.JSON(http.StatusCreated, gin.H{
              "name": input.Name,
            })
          })

          _ = r.Run(":8080")
        }
      `,
      'utf8'
    );

    const getRequest: InvokeRequest = {
      action: 'execute',
      language: 'go',
      filePath: tempFile,
      functionName: '__xgnosis_compat__',
      args: [],
      compatibility: {
        family: 'go-router',
        framework: 'gin',
        request: {
          method: 'GET',
          path: '/users/42',
          query: '?filter=all',
          headers: {
            host: 'compat.test',
          },
        },
      },
    };

    const postRequest: InvokeRequest = {
      action: 'execute',
      language: 'go',
      filePath: tempFile,
      functionName: '__xgnosis_compat__',
      args: [],
      compatibility: {
        family: 'go-router',
        framework: 'gin',
        request: {
          method: 'POST',
          path: '/users',
          headers: {
            host: 'compat.test',
            'content-type': 'application/json',
          },
          body: {
            encoding: 'base64',
            value: Buffer.from(
              JSON.stringify({
                name: 'Ada',
              })
            ).toString('base64'),
          },
        },
      },
    };

    try {
      const getResponse = await invokeSubprocess(getRequest);
      expect(getResponse.status).toBe('ok');
      expect(getResponse.value).toMatchObject({
        matched: true,
        response: {
          status: 200,
          headers: {
            'X-Gin-Compat': '1',
          },
        },
      });
      expect(
        JSON.parse(
          Buffer.from(
            (
              getResponse.value as {
                response: { body: { value: string } };
              }
            ).response.body.value,
            'base64'
          ).toString('utf8')
        )
      ).toEqual({
        id: '42',
        filter: 'all',
      });

      const postResponse = await invokeSubprocess(postRequest);
      expect(postResponse.status).toBe('ok');
      expect(postResponse.value).toMatchObject({
        matched: true,
        response: {
          status: 201,
          headers: {
            'X-Gin-Compat': '1',
          },
        },
      });
      expect(
        JSON.parse(
          Buffer.from(
            (
              postResponse.value as {
                response: { body: { value: string } };
              }
            ).response.body.value,
            'base64'
          ).toString('utf8')
        )
      ).toEqual({
        name: 'Ada',
      });
    } finally {
      removeIfPresent(tempFile);
    }
  }, 20000);

  test('go harness serves chi compatibility requests with stdlib listen interception', async () => {
    const tempFile = path.join(FIXTURES_DIR, '_chi_compat_test.go');
    fs.writeFileSync(
      tempFile,
      `
        package main

        import (
          "encoding/json"
          "net/http"

          "github.com/go-chi/chi/v5"
        )

        func main() {
          r := chi.NewRouter()

          r.Use(func(next http.Handler) http.Handler {
            return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
              w.Header().Set("X-Chi-Compat", "1")
              next.ServeHTTP(w, req)
            })
          })

          r.Get("/users/{id}", func(w http.ResponseWriter, req *http.Request) {
            w.Header().Set("Content-Type", "application/json")
            _ = json.NewEncoder(w).Encode(map[string]string{
              "id": chi.URLParam(req, "id"),
              "marker": req.URL.Query().Get("marker"),
            })
          })

          r.Post("/users", func(w http.ResponseWriter, req *http.Request) {
            var payload map[string]bool
            _ = json.NewDecoder(req.Body).Decode(&payload)
            w.Header().Set("Content-Type", "application/json")
            w.WriteHeader(http.StatusCreated)
            _ = json.NewEncoder(w).Encode(map[string]any{
              "echoed": payload,
            })
          })

          _ = http.ListenAndServe(":3333", r)
        }
      `,
      'utf8'
    );

    const getRequest: InvokeRequest = {
      action: 'execute',
      language: 'go',
      filePath: tempFile,
      functionName: '__xgnosis_compat__',
      args: [],
      compatibility: {
        family: 'go-router',
        framework: 'chi',
        request: {
          method: 'GET',
          path: '/users/42',
          query: '?marker=1',
          headers: {
            host: 'compat.test',
          },
        },
      },
    };

    const postRequest: InvokeRequest = {
      action: 'execute',
      language: 'go',
      filePath: tempFile,
      functionName: '__xgnosis_compat__',
      args: [],
      compatibility: {
        family: 'go-router',
        framework: 'chi',
        request: {
          method: 'POST',
          path: '/users',
          headers: {
            host: 'compat.test',
            'content-type': 'application/json',
          },
          body: {
            encoding: 'base64',
            value: Buffer.from(JSON.stringify({ ok: true })).toString('base64'),
          },
        },
      },
    };

    try {
      const getResponse = await invokeSubprocess(getRequest);
      expect(getResponse.status).toBe('ok');
      expect(getResponse.value).toMatchObject({
        matched: true,
        response: {
          status: 200,
          headers: {
            'X-Chi-Compat': '1',
          },
        },
      });
      expect(
        JSON.parse(
          Buffer.from(
            (
              getResponse.value as {
                response: { body: { value: string } };
              }
            ).response.body.value,
            'base64'
          ).toString('utf8')
        )
      ).toEqual({
        id: '42',
        marker: '1',
      });

      const postResponse = await invokeSubprocess(postRequest);
      expect(postResponse.status).toBe('ok');
      expect(postResponse.value).toMatchObject({
        matched: true,
        response: {
          status: 201,
          headers: {
            'X-Chi-Compat': '1',
          },
        },
      });
      expect(
        JSON.parse(
          Buffer.from(
            (
              postResponse.value as {
                response: { body: { value: string } };
              }
            ).response.body.value,
            'base64'
          ).toString('utf8')
        )
      ).toEqual({
        echoed: { ok: true },
      });
    } finally {
      removeIfPresent(tempFile);
    }
  }, 20000);

  test('go harness serves Fiber compatibility requests with middleware and body parsing', async () => {
    const tempFile = path.join(FIXTURES_DIR, '_fiber_compat_test.go');
    fs.writeFileSync(
      tempFile,
      `
        package main

        import "github.com/gofiber/fiber/v2"

        func main() {
          app := fiber.New()

          app.Use(func(c *fiber.Ctx) error {
            c.Set("X-Fiber-Compat", "1")
            return c.Next()
          })

          app.Get("/users/:id", func(c *fiber.Ctx) error {
            return c.JSON(fiber.Map{
              "id": c.Params("id"),
              "marker": c.Query("marker"),
            })
          })

          app.Post("/users", func(c *fiber.Ctx) error {
            var payload map[string]bool
            _ = c.BodyParser(&payload)
            return c.Status(201).JSON(fiber.Map{
              "echoed": payload,
            })
          })

          _ = app.Listen(":4000")
        }
      `,
      'utf8'
    );

    const getRequest: InvokeRequest = {
      action: 'execute',
      language: 'go',
      filePath: tempFile,
      functionName: '__xgnosis_compat__',
      args: [],
      compatibility: {
        family: 'go-router',
        framework: 'fiber',
        request: {
          method: 'GET',
          path: '/users/42',
          query: '?marker=1',
          headers: {
            host: 'compat.test',
          },
        },
      },
    };

    const postRequest: InvokeRequest = {
      action: 'execute',
      language: 'go',
      filePath: tempFile,
      functionName: '__xgnosis_compat__',
      args: [],
      compatibility: {
        family: 'go-router',
        framework: 'fiber',
        request: {
          method: 'POST',
          path: '/users',
          headers: {
            host: 'compat.test',
            'content-type': 'application/json',
          },
          body: {
            encoding: 'base64',
            value: Buffer.from(JSON.stringify({ ok: true })).toString('base64'),
          },
        },
      },
    };

    try {
      const getResponse = await invokeSubprocess(getRequest);
      expect(getResponse.status).toBe('ok');
      expect(getResponse.value).toMatchObject({
        matched: true,
        response: {
          status: 200,
          headers: {
            'X-Fiber-Compat': '1',
          },
        },
      });
      expect(
        JSON.parse(
          Buffer.from(
            (
              getResponse.value as {
                response: { body: { value: string } };
              }
            ).response.body.value,
            'base64'
          ).toString('utf8')
        )
      ).toEqual({
        id: '42',
        marker: '1',
      });

      const postResponse = await invokeSubprocess(postRequest);
      expect(postResponse.status).toBe('ok');
      expect(postResponse.value).toMatchObject({
        matched: true,
        response: {
          status: 201,
          headers: {
            'X-Fiber-Compat': '1',
          },
        },
      });
      expect(
        JSON.parse(
          Buffer.from(
            (
              postResponse.value as {
                response: { body: { value: string } };
              }
            ).response.body.value,
            'base64'
          ).toString('utf8')
        )
      ).toEqual({
        echoed: { ok: true },
      });
    } finally {
      removeIfPresent(tempFile);
    }
  }, 20000);

  test('go harness serves Echo compatibility requests with middleware and binding', async () => {
    const tempFile = path.join(FIXTURES_DIR, '_echo_compat_test.go');
    fs.writeFileSync(
      tempFile,
      `
        package main

        import (
          "net/http"

          "github.com/labstack/echo/v4"
        )

        func main() {
          e := echo.New()

          e.Use(func(next echo.HandlerFunc) echo.HandlerFunc {
            return func(c echo.Context) error {
              c.Response().Header().Set("X-Echo-Compat", "1")
              return next(c)
            }
          })

          e.GET("/users/:id", func(c echo.Context) error {
            return c.JSON(http.StatusOK, map[string]string{
              "id": c.Param("id"),
              "marker": c.QueryParam("marker"),
            })
          })

          e.POST("/users", func(c echo.Context) error {
            var payload map[string]bool
            _ = c.Bind(&payload)
            return c.JSON(http.StatusCreated, map[string]any{
              "echoed": payload,
            })
          })

          _ = e.Start(":1323")
        }
      `,
      'utf8'
    );

    const getRequest: InvokeRequest = {
      action: 'execute',
      language: 'go',
      filePath: tempFile,
      functionName: '__xgnosis_compat__',
      args: [],
      compatibility: {
        family: 'go-router',
        framework: 'echo',
        request: {
          method: 'GET',
          path: '/users/42',
          query: '?marker=1',
          headers: {
            host: 'compat.test',
          },
        },
      },
    };

    const postRequest: InvokeRequest = {
      action: 'execute',
      language: 'go',
      filePath: tempFile,
      functionName: '__xgnosis_compat__',
      args: [],
      compatibility: {
        family: 'go-router',
        framework: 'echo',
        request: {
          method: 'POST',
          path: '/users',
          headers: {
            host: 'compat.test',
            'content-type': 'application/json',
          },
          body: {
            encoding: 'base64',
            value: Buffer.from(JSON.stringify({ ok: true })).toString('base64'),
          },
        },
      },
    };

    try {
      const getResponse = await invokeSubprocess(getRequest);
      expect(getResponse.status).toBe('ok');
      expect(getResponse.value).toMatchObject({
        matched: true,
        response: {
          status: 200,
          headers: {
            'X-Echo-Compat': '1',
          },
        },
      });
      expect(
        JSON.parse(
          Buffer.from(
            (
              getResponse.value as {
                response: { body: { value: string } };
              }
            ).response.body.value,
            'base64'
          ).toString('utf8')
        )
      ).toEqual({
        id: '42',
        marker: '1',
      });

      const postResponse = await invokeSubprocess(postRequest);
      expect(postResponse.status).toBe('ok');
      expect(postResponse.value).toMatchObject({
        matched: true,
        response: {
          status: 201,
          headers: {
            'X-Echo-Compat': '1',
          },
        },
      });
      expect(
        JSON.parse(
          Buffer.from(
            (
              postResponse.value as {
                response: { body: { value: string } };
              }
            ).response.body.value,
            'base64'
          ).toString('utf8')
        )
      ).toEqual({
        echoed: { ok: true },
      });
    } finally {
      removeIfPresent(tempFile);
    }
  }, 20000);

  test('python harness serves flask compatibility requests with middleware and JSON binding', async () => {
    const tempFile = path.join(FIXTURES_DIR, '_flask_compat_test.py');
    fs.writeFileSync(
      tempFile,
      `
from flask import Flask, jsonify, request

app = Flask(__name__)


@app.before_request
def attach_marker():
    request.compat_marker = request.headers.get('x-compat-marker', 'missing')


@app.get('/users/<user_id>')
def get_user(user_id):
    return jsonify(
        {
            'id': user_id,
            'filter': request.args.get('filter'),
            'marker': request.compat_marker,
        }
    )


@app.post('/users')
def create_user():
    payload = request.get_json() or {}
    return jsonify({'name': payload.get('name')}), 201, {'X-Flask-Compat': '1'}


app.run(port=5000)
      `,
      'utf8'
    );

    const getRequest: InvokeRequest = {
      action: 'execute',
      language: 'python',
      filePath: tempFile,
      functionName: '__xgnosis_compat__',
      args: [],
      compatibility: {
        family: 'python-service',
        framework: 'flask',
        request: {
          method: 'GET',
          path: '/users/42',
          query: '?filter=all',
          headers: {
            host: 'compat.test',
            'x-compat-marker': '1',
          },
        },
      },
    };

    const postRequest: InvokeRequest = {
      action: 'execute',
      language: 'python',
      filePath: tempFile,
      functionName: '__xgnosis_compat__',
      args: [],
      compatibility: {
        family: 'python-service',
        framework: 'flask',
        request: {
          method: 'POST',
          path: '/users',
          headers: {
            host: 'compat.test',
            'content-type': 'application/json',
          },
          body: {
            encoding: 'base64',
            value: Buffer.from(
              JSON.stringify({
                name: 'Ada',
              })
            ).toString('base64'),
          },
        },
      },
    };

    try {
      const getResponse = await invokeSubprocess(getRequest);
      expect(getResponse.status).toBe('ok');
      expect(getResponse.value).toMatchObject({
        matched: true,
        response: {
          status: 200,
          headers: {
            'Content-Type': 'application/json; charset=utf-8',
          },
        },
      });
      expect(
        JSON.parse(
          Buffer.from(
            (
              getResponse.value as {
                response: { body: { value: string } };
              }
            ).response.body.value,
            'base64'
          ).toString('utf8')
        )
      ).toEqual({
        id: '42',
        filter: 'all',
        marker: '1',
      });

      const postResponse = await invokeSubprocess(postRequest);
      expect(postResponse.status).toBe('ok');
      expect(postResponse.value).toMatchObject({
        matched: true,
        response: {
          status: 201,
          headers: {
            'X-Flask-Compat': '1',
            'Content-Type': 'application/json; charset=utf-8',
          },
        },
      });
      expect(
        JSON.parse(
          Buffer.from(
            (
              postResponse.value as {
                response: { body: { value: string } };
              }
            ).response.body.value,
            'base64'
          ).toString('utf8')
        )
      ).toEqual({
        name: 'Ada',
      });
    } finally {
      removeIfPresent(tempFile);
    }
  }, 10000);

  test('python harness serves FastAPI compatibility requests with async middleware and model binding', async () => {
    const tempFile = path.join(FIXTURES_DIR, '_fastapi_compat_test.py');
    fs.writeFileSync(
      tempFile,
      `
from fastapi import FastAPI, Query, Request
from pydantic import BaseModel

app = FastAPI()


class CreateUserInput(BaseModel):
    name: str


@app.middleware('http')
async def attach_marker(request: Request, call_next):
    response = await call_next(request)
    response.headers['X-FastAPI-Compat'] = request.headers.get('x-compat-marker', 'missing')
    return response


@app.get('/users/{user_id}')
async def get_user(user_id: int, filter: str = Query('all')):
    return {
        'id': user_id,
        'filter': filter,
    }


@app.post('/users', status_code=201)
async def create_user(payload: CreateUserInput):
    return payload
      `,
      'utf8'
    );

    const getRequest: InvokeRequest = {
      action: 'execute',
      language: 'python',
      filePath: tempFile,
      functionName: '__xgnosis_compat__',
      args: [],
      compatibility: {
        family: 'python-service',
        framework: 'fastapi',
        request: {
          method: 'GET',
          path: '/users/42',
          query: '?filter=all',
          headers: {
            host: 'compat.test',
            'x-compat-marker': '1',
          },
        },
      },
    };

    const postRequest: InvokeRequest = {
      action: 'execute',
      language: 'python',
      filePath: tempFile,
      functionName: '__xgnosis_compat__',
      args: [],
      compatibility: {
        family: 'python-service',
        framework: 'fastapi',
        request: {
          method: 'POST',
          path: '/users',
          headers: {
            host: 'compat.test',
            'content-type': 'application/json',
            'x-compat-marker': '1',
          },
          body: {
            encoding: 'base64',
            value: Buffer.from(
              JSON.stringify({
                name: 'Ada',
              })
            ).toString('base64'),
          },
        },
      },
    };

    try {
      const getResponse = await invokeSubprocess(getRequest);
      expect(getResponse.status).toBe('ok');
      expect(getResponse.value).toMatchObject({
        matched: true,
        response: {
          status: 200,
          headers: {
            'X-FastAPI-Compat': '1',
            'Content-Type': 'application/json; charset=utf-8',
          },
        },
      });
      expect(
        JSON.parse(
          Buffer.from(
            (
              getResponse.value as {
                response: { body: { value: string } };
              }
            ).response.body.value,
            'base64'
          ).toString('utf8')
        )
      ).toEqual({
        id: 42,
        filter: 'all',
      });

      const postResponse = await invokeSubprocess(postRequest);
      expect(postResponse.status).toBe('ok');
      expect(postResponse.value).toMatchObject({
        matched: true,
        response: {
          status: 201,
          headers: {
            'X-FastAPI-Compat': '1',
            'Content-Type': 'application/json; charset=utf-8',
          },
        },
      });
      expect(
        JSON.parse(
          Buffer.from(
            (
              postResponse.value as {
                response: { body: { value: string } };
              }
            ).response.body.value,
            'base64'
          ).toString('utf8')
        )
      ).toEqual({
        name: 'Ada',
      });
    } finally {
      removeIfPresent(tempFile);
    }
  }, 10000);

  test('ruby harness serves sinatra compatibility requests with middleware and JSON binding', async () => {
    const tempFile = path.join(FIXTURES_DIR, '_sinatra_compat_test.rb');
    fs.writeFileSync(
      tempFile,
      `
require 'json'
require 'sinatra'

set :port, 4567

before do
  headers 'X-Sinatra-Compat' => '1'
end

get '/users/:id' do
  content_type :json
  JSON.generate({
    id: params['id'],
    filter: params['filter'],
  })
end

post '/users' do
  payload = JSON.parse(request.body.read)
  status 201
  content_type :json
  JSON.generate({
    name: payload['name'],
  })
end
      `,
      'utf8'
    );

    const getRequest: InvokeRequest = {
      action: 'execute',
      language: 'ruby',
      filePath: tempFile,
      functionName: '__xgnosis_compat__',
      args: [],
      compatibility: {
        family: 'ruby-service',
        framework: 'sinatra',
        request: {
          method: 'GET',
          path: '/users/42',
          query: '?filter=all',
          headers: {
            host: 'compat.test',
          },
        },
      },
    };

    const postRequest: InvokeRequest = {
      action: 'execute',
      language: 'ruby',
      filePath: tempFile,
      functionName: '__xgnosis_compat__',
      args: [],
      compatibility: {
        family: 'ruby-service',
        framework: 'sinatra',
        request: {
          method: 'POST',
          path: '/users',
          headers: {
            host: 'compat.test',
            'content-type': 'application/json',
          },
          body: {
            encoding: 'base64',
            value: Buffer.from(
              JSON.stringify({
                name: 'Ada',
              })
            ).toString('base64'),
          },
        },
      },
    };

    try {
      const getResponse = await invokeSubprocess(getRequest);
      expect(getResponse.status).toBe('ok');
      expect(getResponse.value).toMatchObject({
        matched: true,
        response: {
          status: 200,
          headers: {
            'Content-Type': 'application/json; charset=utf-8',
            'X-Sinatra-Compat': '1',
          },
        },
      });
      expect(
        JSON.parse(
          Buffer.from(
            (
              getResponse.value as {
                response: { body: { value: string } };
              }
            ).response.body.value,
            'base64'
          ).toString('utf8')
        )
      ).toEqual({
        id: '42',
        filter: 'all',
      });

      const postResponse = await invokeSubprocess(postRequest);
      expect(postResponse.status).toBe('ok');
      expect(postResponse.value).toMatchObject({
        matched: true,
        response: {
          status: 201,
          headers: {
            'Content-Type': 'application/json; charset=utf-8',
            'X-Sinatra-Compat': '1',
          },
        },
      });
      expect(
        JSON.parse(
          Buffer.from(
            (
              postResponse.value as {
                response: { body: { value: string } };
              }
            ).response.body.value,
            'base64'
          ).toString('utf8')
        )
      ).toEqual({
        name: 'Ada',
      });
    } finally {
      removeIfPresent(tempFile);
    }
  }, 10000);

  test('java harness serves spring compatibility requests with path, query, and request-body binding', async () => {
    const tempFile = path.join(FIXTURES_DIR, 'SpringCompatController.java');
    fs.writeFileSync(
      tempFile,
      `
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api")
public class SpringCompatController {
  @GetMapping("/users/{id}")
  public String getUser(
      @PathVariable("id") String id,
      @RequestParam("filter") String filter
  ) {
    return "{\\"id\\":\\"" + id + "\\",\\"filter\\":\\"" + filter + "\\"}";
  }

  @PostMapping("/users")
  @ResponseStatus(HttpStatus.CREATED)
  public String createUser(@RequestBody String body) {
    return body;
  }
}
      `,
      'utf8'
    );

    const getRequest: InvokeRequest = {
      action: 'execute',
      language: 'java',
      filePath: tempFile,
      functionName: '__xgnosis_compat__',
      args: [],
      compatibility: {
        family: 'java-service',
        framework: 'spring',
        request: {
          method: 'GET',
          path: '/api/users/42',
          query: '?filter=all',
          headers: {
            host: 'compat.test',
          },
        },
      },
    };

    const postRequest: InvokeRequest = {
      action: 'execute',
      language: 'java',
      filePath: tempFile,
      functionName: '__xgnosis_compat__',
      args: [],
      compatibility: {
        family: 'java-service',
        framework: 'spring',
        request: {
          method: 'POST',
          path: '/api/users',
          headers: {
            host: 'compat.test',
            'content-type': 'application/json',
          },
          body: {
            encoding: 'base64',
            value: Buffer.from(
              JSON.stringify({
                name: 'Ada',
              })
            ).toString('base64'),
          },
        },
      },
    };

    try {
      const getResponse = await invokeSubprocess(getRequest);
      expect(getResponse.status).toBe('ok');
      expect(getResponse.value).toMatchObject({
        matched: true,
        response: {
          status: 200,
          headers: {
            'Content-Type': 'application/json; charset=utf-8',
          },
        },
      });
      expect(
        JSON.parse(
          Buffer.from(
            (
              getResponse.value as {
                response: { body: { value: string } };
              }
            ).response.body.value,
            'base64'
          ).toString('utf8')
        )
      ).toEqual({
        id: '42',
        filter: 'all',
      });

      const postResponse = await invokeSubprocess(postRequest);
      expect(postResponse.status).toBe('ok');
      expect(postResponse.value).toMatchObject({
        matched: true,
        response: {
          status: 201,
          headers: {
            'Content-Type': 'application/json; charset=utf-8',
          },
        },
      });
      expect(
        JSON.parse(
          Buffer.from(
            (
              postResponse.value as {
                response: { body: { value: string } };
              }
            ).response.body.value,
            'base64'
          ).toString('utf8')
        )
      ).toEqual({
        name: 'Ada',
      });
    } finally {
      removeIfPresent(tempFile);
    }
  }, 20000);

  for (const fixture of SCRIPT_SERVICE_FIXTURES) {
    if (!fixture.isAvailable()) {
      continue;
    }

    test(
      `${fixture.language} script-service serves standard request envelopes`,
      async () => {
        const tempFile = makeScriptServiceTempFilePath(fixture);
        fs.writeFileSync(
          tempFile,
          normalizeFixtureSource(fixture.sourceText),
          'utf8'
        );

        const request: InvokeRequest = {
          action: 'execute',
          language: fixture.language,
          filePath: tempFile,
          functionName: '__xgnosis_compat__',
          args: [],
          compatibility: {
            family: 'script-service',
            framework: 'standard',
            request: {
              method: 'GET',
              path: '/standard',
              headers: {
                host: 'compat.test',
              },
            },
          },
        };

        try {
          const response = await withSubprocessTimeout(
            fixture.timeoutMs ?? 60000,
            async () => invokeSubprocess(request)
          );

          if (response.status !== 'ok') {
            throw new Error(
              `${fixture.language} script-service failed: ${JSON.stringify(
                response
              )}`
            );
          }

          expect(response.status).toBe('ok');
          expect(response.value).toMatchObject({
            matched: true,
            response: {
              status: 200,
            },
          });
          expect(decodeCompatibilityBody(response.value)).toBe(
            fixture.expectedBody
          );
        } finally {
          removeIfPresent(tempFile);
        }
      },
      fixture.timeoutMs ?? 60000
    );
  }

  test('ping action returns pong', async () => {
    const response = await invokeSubprocess({
      action: 'ping',
      language: 'python',
      filePath: '',
      functionName: '',
      args: [],
    });
    expect(response.status).toBe('ok');
    expect(response.value).toBe('pong');
  }, 10000);
});

// ─── Polyglot Bridge Handler Tests ───────────────────────────────────────────

describe('polyglot bridge handlers', () => {
  test('registers all five handler types', () => {
    const registry = new GnosisRegistry();
    const manifest: PolyglotExecutionManifest = {
      language: 'python',
      file_path: '/test.py',
      entry_function: 'main',
      node_execution_plans: [],
    };

    registerPolyglotBridgeHandlers(registry, manifest);

    expect(registry.hasHandler('PolyglotBridgeEntry')).toBe(true);
    expect(registry.hasHandler('PolyglotBridgeCall')).toBe(true);
    expect(registry.hasHandler('PolyglotBridgeStatement')).toBe(true);
    expect(registry.hasHandler('PolyglotBridgeReturn')).toBe(true);
    expect(registry.hasHandler('PolyglotBridgeJoin')).toBe(true);
  });

  test('entry handler initializes state', async () => {
    const registry = new GnosisRegistry();
    const manifest: PolyglotExecutionManifest = {
      language: 'python',
      file_path: '/test.py',
      entry_function: 'main',
      node_execution_plans: [],
    };

    registerPolyglotBridgeHandlers(registry, manifest);

    const handler = registry.getHandler('PolyglotBridgeEntry')!;
    const result = await handler({ x: 42 }, {});

    expect(result).toMatchObject({
      input: { x: 42 },
      locals: {},
      language: 'python',
      filePath: '/test.py',
    });
  });

  test('return handler extracts last value', async () => {
    const registry = new GnosisRegistry();
    const manifest: PolyglotExecutionManifest = {
      language: 'python',
      file_path: '/test.py',
      entry_function: 'main',
      node_execution_plans: [],
    };

    registerPolyglotBridgeHandlers(registry, manifest);

    const handler = registry.getHandler('PolyglotBridgeReturn')!;
    const result = await handler(
      {
        input: {},
        locals: { __last__: 'final_value' },
        language: 'python',
        filePath: '/test.py',
      },
      {},
      { nodeId: 'return_0' }
    );

    expect(result).toBe('final_value');
  });
});

// ─── Multi-Language Bridge Tests ─────────────────────────────────────────────

describe('multi-language bridge', () => {
  test('buildMultiLanguageManifest extracts overrides from node properties', () => {
    const nodes = new Map<string, ASTNode>();
    nodes.set('ml_predict', {
      id: 'ml_predict',
      labels: ['Statement', 'PolyglotBridgeCall'],
      properties: { language: 'python', file: 'ml.py', callee: 'predict' },
    });
    nodes.set('api_call', {
      id: 'api_call',
      labels: ['Statement', 'PolyglotBridgeCall'],
      properties: { language: 'go', file: 'api.go', callee: 'fetch' },
    });
    nodes.set('entry', {
      id: 'entry',
      labels: ['Entry'],
      properties: {},
    });

    const ast: GraphAST = { nodes, edges: [] };
    const manifest = buildMultiLanguageManifest(
      ast,
      'javascript',
      '/default'
    );

    expect(manifest.defaultLanguage).toBe('javascript');
    expect(manifest.nodeOverrides.size).toBe(2);
    expect(manifest.nodeOverrides.get('ml_predict')).toMatchObject({
      language: 'python',
      filePath: 'ml.py',
    });
    expect(manifest.nodeOverrides.get('api_call')).toMatchObject({
      language: 'go',
      filePath: 'api.go',
    });
  });
});

// ─── Cross-Language Trace Tests ──────────────────────────────────────────────

describe('cross-language execution trace', () => {
  test('builds backward trace from failed node', () => {
    const nodes = new Map<string, ASTNode>();
    nodes.set('entry', {
      id: 'entry',
      labels: ['Entry', 'PolyglotBridgeEntry'],
      properties: { name: 'pipeline', language: 'javascript' },
    });
    nodes.set('parse', {
      id: 'parse',
      labels: ['Statement', 'PolyglotBridgeCall'],
      properties: { language: 'rust', callee: 'parse_input' },
    });
    nodes.set('fetch', {
      id: 'fetch',
      labels: ['Statement', 'PolyglotBridgeCall'],
      properties: { language: 'go', callee: 'fetch_data' },
    });
    nodes.set('predict', {
      id: 'predict',
      labels: ['Statement', 'PolyglotBridgeCall'],
      properties: { language: 'python', callee: 'ml_predict' },
    });

    const edges: ASTEdge[] = [
      {
        sourceIds: ['entry'],
        targetIds: ['parse'],
        type: 'PROCESS',
        properties: {},
      },
      {
        sourceIds: ['parse'],
        targetIds: ['fetch'],
        type: 'PROCESS',
        properties: {},
      },
      {
        sourceIds: ['fetch'],
        targetIds: ['predict'],
        type: 'PROCESS',
        properties: {},
      },
    ];

    const ast: GraphAST = { nodes, edges };
    const manifest: PolyglotExecutionManifest = {
      language: 'javascript',
      file_path: '/pipeline',
      entry_function: 'pipeline',
      node_execution_plans: [
        {
          node_id: 'entry',
          source_range: { start_byte: 0, end_byte: 10 },
          kind: 'entry',
        },
        {
          node_id: 'parse',
          source_range: { start_byte: 10, end_byte: 50 },
          kind: 'call',
          callee: 'parse_input',
        },
        {
          node_id: 'fetch',
          source_range: { start_byte: 50, end_byte: 90 },
          kind: 'call',
          callee: 'fetch_data',
        },
        {
          node_id: 'predict',
          source_range: { start_byte: 90, end_byte: 130 },
          kind: 'call',
          callee: 'ml_predict',
        },
      ],
    };

    const mappedError = {
      language: 'python',
      filePath: 'ml.py',
      functionName: 'ml_predict',
      message: 'ValueError: could not convert string to float',
      line: 42,
      traceback: 'Traceback...',
    };

    const trace = buildExecutionTrace(ast, 'predict', mappedError, manifest);

    expect(trace.frames.length).toBe(4);
    expect(trace.failedNode).toBe('predict');
    expect(trace.analysis.languages).toContain('python');
    expect(trace.analysis.languages).toContain('go');
    expect(trace.analysis.languages).toContain('rust');
    expect(trace.analysis.errorKind).toBe('semantic');
    expect(trace.frames[3]!.status).toBe('error');
    expect(trace.frames[3]!.line).toBe(42);
  });

  test('formatExecutionTrace produces readable output', () => {
    const trace = {
      frames: [
        {
          nodeId: 'entry',
          labels: ['Entry'],
          language: 'javascript',
          filePath: '/p',
          functionName: 'pipeline',
          status: 'completed' as const,
          incomingEdgeType: undefined,
        },
        {
          nodeId: 'predict',
          labels: ['Statement'],
          language: 'python',
          filePath: 'ml.py',
          functionName: 'ml_predict',
          status: 'error' as const,
          line: 42,
          incomingEdgeType: 'PROCESS',
        },
      ],
      failedNode: 'predict',
      mappedError: {
        language: 'python',
        filePath: 'ml.py',
        functionName: 'ml_predict',
        message: 'ValueError',
        line: 42,
        traceback: '',
      },
      analysis: {
        pathLength: 2,
        languages: ['javascript', 'python'],
        errorKind: 'semantic' as const,
        edgeTypes: ['PROCESS'],
      },
    };

    const formatted = formatExecutionTrace(trace);
    expect(formatted).toContain('predict');
    expect(formatted).toContain('python');
    expect(formatted).toContain('42');
    expect(formatted).toContain('semantic');
  });
});

// ─── Compose Tests ───────────────────────────────────────────────────────────

describe('polyglot compose', () => {
  test('composes two analysis results into a topology', () => {
    const pyAnalysis = {
      functions: [
        {
          functionName: 'process',
          ast: { nodes: new Map(), edges: [] } as GraphAST,
          ggSource: '(entry_0: Entry)',
        },
      ],
      language: 'python',
      errors: [],
    };

    const goAnalysis = {
      functions: [
        {
          functionName: 'serve',
          ast: { nodes: new Map(), edges: [] } as GraphAST,
          ggSource: '(entry_0: Entry)',
        },
      ],
      language: 'go',
      errors: [],
    };

    const result = compose([pyAnalysis, goAnalysis], ['/app.py', '/server.go']);

    expect(result.languages).toContain('python');
    expect(result.languages).toContain('go');
    expect(result.functions.length).toBe(2);
    expect(result.ggSource).toContain('PolyglotBridgeCall');
    expect(result.ggSource).toContain("language='python'");
    expect(result.ggSource).toContain("language='go'");
    expect(result.connections.length).toBeGreaterThanOrEqual(1);
  });

  test('infers pipeline connections from function names', () => {
    const analysis = {
      functions: [
        {
          functionName: 'parse_input',
          ast: { nodes: new Map(), edges: [] } as GraphAST,
          ggSource: '',
        },
        {
          functionName: 'transform_data',
          ast: { nodes: new Map(), edges: [] } as GraphAST,
          ggSource: '',
        },
        {
          functionName: 'output_result',
          ast: { nodes: new Map(), edges: [] } as GraphAST,
          ggSource: '',
        },
      ],
      language: 'python',
      errors: [],
    };

    const result = compose([analysis, analysis], ['/a.py', '/b.py']);
    // Should detect pipeline pattern: parse → transform → output.
    expect(result.connections.length).toBeGreaterThanOrEqual(1);
  });
});

// ─── Translate Tests ─────────────────────────────────────────────────────────

describe('polyglot translate', () => {
  test('generates skeleton in target language', () => {
    const analysis = {
      functions: [
        {
          functionName: 'process_data',
          ast: { nodes: new Map(), edges: [] } as GraphAST,
          ggSource: '(entry_0: Entry)',
        },
      ],
      language: 'python',
      errors: [],
    };

    const result = translate(analysis, '/app.py', 'go');

    expect(result.files.length).toBe(1);
    expect(result.files[0]!.fileName).toContain('.go');
    expect(result.files[0]!.source).toContain('func');
    expect(result.files[0]!.source).toContain('ProcessData');
  });

  test('translates to rust', () => {
    const analysis = {
      functions: [
        {
          functionName: 'compute',
          ast: { nodes: new Map(), edges: [] } as GraphAST,
          ggSource: '(entry_0: Entry)',
        },
      ],
      language: 'python',
      errors: [],
    };

    const result = translate(analysis, '/app.py', 'rust');
    expect(result.files[0]!.fileName).toContain('.rs');
    expect(result.files[0]!.source).toContain('pub fn');
  });
});

// ─── Best Language Tests ─────────────────────────────────────────────────────

describe('best language fitness', () => {
  test('ranks languages by fitness for compute-heavy function', () => {
    const func = {
      filePath: '/compute.py',
      language: 'python',
      name: 'matrix_multiply',
      params: [],
      callees: ['numpy.dot', 'reshape'],
      ast: { nodes: new Map(), edges: [] } as GraphAST,
      ggSource: '',
      nodeCount: 15,
      edgeTypes: ['PROCESS', 'FORK', 'FOLD'],
    };

    const fitness = findBestLanguage(func);
    expect(fitness.length).toBeGreaterThan(5);
    // Compiled languages should rank higher for compute.
    const topThree = fitness.slice(0, 3).map((f) => f.language);
    expect(
      topThree.includes('rust') ||
        topThree.includes('go') ||
        topThree.includes('c')
    ).toBe(true);
  });

  test('computeBetaCost calculates correctly', () => {
    const nodes = new Map<string, ASTNode>();
    nodes.set('a', { id: 'a', labels: ['Entry'], properties: {} });
    nodes.set('b', { id: 'b', labels: ['Statement'], properties: {} });
    nodes.set('c', { id: 'c', labels: ['Statement'], properties: {} });
    nodes.set('d', { id: 'd', labels: ['Return'], properties: {} });

    const edges: ASTEdge[] = [
      { sourceIds: ['a'], targetIds: ['b', 'c'], type: 'FORK', properties: {} },
      { sourceIds: ['b', 'c'], targetIds: ['d'], type: 'FOLD', properties: {} },
    ];

    const func = {
      filePath: '/t.py',
      language: 'python',
      name: 'test',
      params: [],
      callees: [],
      ast: { nodes, edges } as GraphAST,
      ggSource: '',
      nodeCount: 4,
      edgeTypes: ['FORK', 'FOLD'],
    };

    const cost = computeBetaCost(func);
    expect(cost.totalBeta).toBe(2); // FORK(2 targets)=1, FOLD(2 sources)=1
    expect(cost.leastBuleScore).toBeGreaterThan(0);
  });
});

// ─── Topo-Race Tests ─────────────────────────────────────────────────────────

describe('topo-race topology generation', () => {
  test('generates valid RACE topology', () => {
    const topology = generateTopoRaceTopology('fibonacci', '/fib.py', [
      'python',
      'go',
      'rust',
    ]);

    expect(topology).toContain('RACE');
    expect(topology).toContain('FOLD');
    expect(topology).toContain("language='python'");
    expect(topology).toContain("language='go'");
    expect(topology).toContain("language='rust'");
    expect(topology).toContain('fibonacci_python');
    expect(topology).toContain('fibonacci_go');
    expect(topology).toContain('fibonacci_rust');
  });
});

// ─── Scaffold Tests ──────────────────────────────────────────────────────────

describe('topology scaffolder', () => {
  test('generates code for multiple languages', () => {
    const nodes = new Map<string, ASTNode>();
    nodes.set('ml_node', {
      id: 'ml_node',
      labels: ['Statement', 'PolyglotBridgeCall'],
      properties: { callee: 'predict' },
    });
    nodes.set('api_node', {
      id: 'api_node',
      labels: ['Statement', 'PolyglotBridgeCall'],
      properties: { callee: 'serve' },
    });

    const ast: GraphAST = { nodes, edges: [] };

    const result = scaffoldTopology(ast, [
      { nodeId: 'ml_node', language: 'python', functionName: 'predict' },
      { nodeId: 'api_node', language: 'go', functionName: 'serve' },
    ]);

    expect(result.files.length).toBe(2);
    expect(result.files[0]!.language).toBe('python');
    expect(result.files[0]!.source).toContain('def predict');
    expect(result.files[1]!.language).toBe('go');
    expect(result.files[1]!.source).toContain('func Serve');
  });
});

// ─── Strategy Memory Tests ───────────────────────────────────────────────────

describe('strategy memory CRDT', () => {
  test('records observations and computes scores', () => {
    const memory = new PolyglotStrategyMemory({
      decay: { halfLife: 5, maxObservations: 20, maxAgeDays: 7 },
    });

    memory.record({
      functionName: 'fib',
      language: 'python',
      filePath: '/f.py',
      durationMs: 100,
      success: true,
      timestamp: new Date().toISOString(),
    });
    memory.record({
      functionName: 'fib',
      language: 'rust',
      filePath: '/f.rs',
      durationMs: 5,
      success: true,
      timestamp: new Date().toISOString(),
    });

    const pyScore = memory.score('fib', 'python');
    const rsScore = memory.score('fib', 'rust');

    expect(pyScore).not.toBeNull();
    expect(rsScore).not.toBeNull();
    expect(rsScore!.fitnessScore).toBeGreaterThan(pyScore!.fitnessScore);
  });

  test('ranks languages correctly', () => {
    const memory = new PolyglotStrategyMemory();

    for (let i = 0; i < 5; i++) {
      memory.record({
        functionName: 'sort',
        language: 'python',
        filePath: '/s.py',
        durationMs: 200,
        success: true,
        timestamp: new Date().toISOString(),
      });
      memory.record({
        functionName: 'sort',
        language: 'go',
        filePath: '/s.go',
        durationMs: 10,
        success: true,
        timestamp: new Date().toISOString(),
      });
      memory.record({
        functionName: 'sort',
        language: 'rust',
        filePath: '/s.rs',
        durationMs: 3,
        success: true,
        timestamp: new Date().toISOString(),
      });
    }

    const ranked = memory.rank('sort');
    expect(ranked.length).toBe(3);
    expect(ranked[0]!.language).toBe('rust'); // fastest
    expect(ranked[2]!.language).toBe('python'); // slowest
  });

  test('recommend returns best language', () => {
    const memory = new PolyglotStrategyMemory();

    memory.record({
      functionName: 'calc',
      language: 'python',
      filePath: '/c.py',
      durationMs: 500,
      success: true,
      timestamp: new Date().toISOString(),
    });
    memory.record({
      functionName: 'calc',
      language: 'go',
      filePath: '/c.go',
      durationMs: 20,
      success: true,
      timestamp: new Date().toISOString(),
    });

    expect(memory.recommend('calc')).toBe('go');
  });

  test('serialize and restore preserves state', () => {
    const memory = new PolyglotStrategyMemory();
    memory.record({
      functionName: 'f',
      language: 'python',
      filePath: '/f.py',
      durationMs: 100,
      success: true,
      timestamp: new Date().toISOString(),
    });

    const serialized = memory.serialize();
    const restored = new PolyglotStrategyMemory();
    restored.restore(serialized);

    expect(restored.score('f', 'python')).not.toBeNull();
    expect(restored.score('f', 'python')!.observations).toBe(1);
  });

  test('penalizes error-prone languages', () => {
    const memory = new PolyglotStrategyMemory();

    // Python succeeds but is slow.
    for (let i = 0; i < 5; i++) {
      memory.record({
        functionName: 'api',
        language: 'python',
        filePath: '/a.py',
        durationMs: 100,
        success: true,
        timestamp: new Date().toISOString(),
      });
    }

    // Go is fast but errors often.
    for (let i = 0; i < 5; i++) {
      memory.record({
        functionName: 'api',
        language: 'go',
        filePath: '/a.go',
        durationMs: 10,
        success: i < 2,
        timestamp: new Date().toISOString(),
      });
    }

    const ranked = memory.rank('api');
    // Python should win because Go has 60% error rate.
    expect(ranked[0]!.language).toBe('python');
  });
});

// ─── Optimizer Tests ─────────────────────────────────────────────────────────

describe('autonomous optimizer', () => {
  test('suggests RACE for slow nodes', () => {
    const nodes = new Map<string, ASTNode>();
    nodes.set('slow', { id: 'slow', labels: ['Statement'], properties: {} });
    const ast: GraphAST = { nodes, edges: [] };

    const profiler = new TopologyProfiler();
    for (let i = 0; i < 10; i++) {
      profiler.record('slow', 'python', 'compute', 1000, false);
    }

    const suggestions = analyzeForOptimizations(ast, profiler, {
      slowThresholdMs: 500,
      minExecutions: 5,
    });

    expect(suggestions.length).toBeGreaterThan(0);
    expect(suggestions[0]!.kind).toBe('race');
    expect(suggestions[0]!.description).toContain('slow');
  });

  test('suggests fallback for error-prone nodes', () => {
    const nodes = new Map<string, ASTNode>();
    nodes.set('flaky', { id: 'flaky', labels: ['Statement'], properties: {} });
    const ast: GraphAST = { nodes, edges: [] };

    const profiler = new TopologyProfiler();
    for (let i = 0; i < 10; i++) {
      profiler.record('flaky', 'python', 'api_call', 50, i < 3); // 70% error rate
    }

    const suggestions = analyzeForOptimizations(ast, profiler, {
      errorRateThreshold: 0.1,
      minExecutions: 5,
    });

    const fallback = suggestions.find((s) => s.kind === 'fallback');
    expect(fallback).toBeDefined();
    expect(fallback!.description).toContain('error rate');
  });
});
