// @a0n/gnosis/src/__tests__/study-participant-link.test.ts
import {
  buildStudyAggregateSidecar,
  buildStudyParticipantIngressRef,
  type StudyParticipantIngressRefV1,
} from '../study-participant-link';

export default function law(): void {
  const agg = buildStudyAggregateSidecar({
    studyId: 's1',
    instrumentVersion: 'v1',
    responseCount: 10,
    completedCount: 7,
    exportedAtIso: '2026-01-01T00:00:00.000Z',
  });
  if (agg.study_id !== 's1' || agg.response_count !== 10) {
    throw new Error('aggregate');
  }

  const ref: StudyParticipantIngressRefV1 = buildStudyParticipantIngressRef({
    snapshot: {
      study_id: 'study-abc',
      preregistration_ref: 'https://osf.io/xyz',
    },
    consentVersion: '2026-04-17',
    instrumentVersion: 'v2',
    inviteTokenHash: 'deadbeef',
  });
  if (ref.schemaVersion !== '1.0.0') throw new Error('schema');
  if (ref.study_id !== 'study-abc') throw new Error('study_id');
  if (ref.consent_version !== '2026-04-17') throw new Error('consent');
  if (ref.instrument_version !== 'v2') throw new Error('instrument');
  if (ref.invite_token_hash !== 'deadbeef') throw new Error('hash');
  if (ref.preregistration_ref !== 'https://osf.io/xyz') throw new Error('prereg');
}
