// @a0n/.ignored_gnosis/forkjoin-app/client/pages/AdminPage.tsx
import React from '@a0n/raect';
import { Guard } from '@a0n/aeon-flags/react';

/**
 * Renders the Admin Page view.
 */
export function AdminPage() {
  const AccessDenied = (
    <div
      className="flex h-screen flex-col items-center justify-center bg-[#0a0a0d] p-8"
      style={{
        background: 'var(--color-bg-root)',
        color: 'var(--color-text-primary)',
        padding: '2rem',
      }}
    >
      <h2 className="text-2xl font-bold mb-4 text-red-500">Access Denied</h2>
      <p style={{ color: 'var(--color-text-secondary)' }}>
        Valid Admin UCAN capability is required to access the AeonDev Admin
        Dashboard.
      </p>
    </div>
  );

  return (
    <Guard flag="admin-dashboard" fallback={AccessDenied}>
      <div
        className="flex min-h-screen flex-col bg-[#0a0a0d] p-8"
        style={{
          background: 'var(--color-bg-root)',
          color: 'var(--color-text-primary)',
          padding: '2rem',
        }}
      >
        <h1
          className="text-3xl font-bold mb-2"
          style={{ marginBottom: '0.5rem' }}
        >
          AeonDev Admin Dashboard
        </h1>
        <p
          className="mb-8"
          style={{ color: 'var(--color-text-secondary)', marginBottom: '2rem' }}
        >
          Operational controls for benchmark orchestration, webhooks, and model
          registry workflows.
        </p>

        <div
          className="grid gap-6 md:grid-cols-2 lg:grid-cols-3"
          style={{
            display: 'grid',
            gridTemplateColumns: 'repeat(auto-fit, minmax(300px, 1fr))',
            gap: '1.5rem',
          }}
        >
          <a href="/admin/benchmarks" className="admin-card" style={cardStyle}>
            <h2
              style={{ color: 'var(--color-accent)', marginBottom: '0.5rem' }}
            >
              Benchmarks
            </h2>
            <p
              style={{
                color: 'var(--color-text-tertiary)',
                fontSize: '0.875rem',
                margin: 0,
              }}
            >
              Trigger and monitor benchmark runs.
            </p>
          </a>
          <a href="/admin/models" className="admin-card" style={cardStyle}>
            <h2
              style={{ color: 'var(--color-accent)', marginBottom: '0.5rem' }}
            >
              Models
            </h2>
            <p
              style={{
                color: 'var(--color-text-tertiary)',
                fontSize: '0.875rem',
                margin: 0,
              }}
            >
              Inspect model availability and deployment state.
            </p>
          </a>
          <a href="/admin/webhooks" className="admin-card" style={cardStyle}>
            <h2
              style={{ color: 'var(--color-accent)', marginBottom: '0.5rem' }}
            >
              Webhooks
            </h2>
            <p
              style={{
                color: 'var(--color-text-tertiary)',
                fontSize: '0.875rem',
                margin: 0,
              }}
            >
              Configure and inspect inbound automation webhooks.
            </p>
          </a>
        </div>
      </div>
    </Guard>
  );
}

const cardStyle = {
  display: 'block',
  padding: '1.5rem',
  borderRadius: '0.75rem',
  border: '1px solid var(--color-bg-tertiary)',
  background: 'var(--color-bg-secondary)',
  textDecoration: 'none',
  transition: 'all 0.2s ease',
};
