// @a0n/.ignored_gnosis/forkjoin-app/client/pages/Docs.tsx
import {
  FOUNDATION_DOCS_TRACK_IDS,
  getDocsTrack,
  type DocsTrackId,
} from '@a0n/shared-utils/docsManifest';
import {
  ArrowLeft,
  BookOpen,
  FileCode,
  RefreshCw,
  Database,
  Network,
  Shield,
  Waves,
  Globe,
  Brain,
} from '@a0n/aeon-icons';
import {
  FOUNDATION_DOCS_HOME_URL,
  FOUNDATION_DOCS_LIBRARY_URL,
  FOUNDATION_DOCS_PRODUCTS_URL,
  FOUNDATION_DOCS_QUICKSTART_URL,
  getFoundationApiTrackUrl,
  getFoundationDocsTrackUrl,
  getFoundationTrackBookUrl,
} from '../docsLinks';
import type { JSX } from '@a0n/raect';

const TRACK_ICONS: Record<DocsTrackId, JSX.Element> = {
  aeon: <RefreshCw size={24} />,
  'aeon-flux': <Waves size={24} />,
  dash: <Database size={24} />,
  aegis: <Shield size={24} />,
  dashrelay: <Network size={24} />,
  edgework: <Globe size={24} />,
  neural: <Brain size={24} />,
};

const launchPads = [
  {
    title: 'Gnosis Church',
    description:
      'Taste the language first: one-minute orientation, real code, quickstart, and Book 153.',
    href: 'https://gnosis.church',
    icon: <BookOpen size={24} />,
  },
  {
    title: 'Quickstart',
    description:
      'Go straight to installation, local setup, and the first working path.',
    href: FOUNDATION_DOCS_QUICKSTART_URL,
    icon: <BookOpen size={24} />,
  },
  {
    title: 'Docs Home',
    description: 'Start from the high-level map of APIs, packages, and guides.',
    href: FOUNDATION_DOCS_HOME_URL,
    icon: <FileCode size={24} />,
  },
  {
    title: 'Products',
    description:
      'See how the technical surfaces connect to the shipped platform.',
    href: FOUNDATION_DOCS_PRODUCTS_URL,
    icon: <Globe size={24} />,
  },
  {
    title: 'Library',
    description:
      'Browse the ebook collection when you want depth instead of a quick answer.',
    href: FOUNDATION_DOCS_LIBRARY_URL,
    icon: <BookOpen size={24} />,
  },
];

const gnosisProcession = [
  {
    title: 'Taste in Church',
    description:
      'Use gnosis.church as the evaluator surface: awe, code taste, quickstart, and the flagship book.',
    href: 'https://gnosis.church',
    label: 'Church',
  },
  {
    title: 'Read the exact docs',
    description:
      'Once the language shape feels right, move into the canonical docs corpus for installation and task answers.',
    href: getFoundationDocsTrackUrl('aeon'),
    label: 'Docs',
  },
  {
    title: 'Start with Book 153',
    description:
      'Use the flagship language book when you want the full rationale instead of just the quickstart path.',
    href: 'https://docs.aeonflux.dev/#/page/ebooks/153-gnosis-topological-programming-language/README',
    label: 'Book 153',
  },
  {
    title: 'Build, deploy, open',
    description:
      'The serious flow is IDE for building, Forge for deploy, AeonOS for open.',
    href: 'https://ide.aeonflux.dev',
    label: 'IDE -> Forge -> AeonOS',
  },
] as const;

/**
 * Renders the Docs view.
 */
export function Docs() {
  const tracks = FOUNDATION_DOCS_TRACK_IDS.flatMap((trackId) => {
    const track = getDocsTrack(trackId);
    return track ? [track] : [];
  });

  return (
    <div className="docs-page">
      <nav className="docs-nav">
        <a href="/" className="back-link">
          <ArrowLeft size={20} />
          <span>Back to Aeon Foundation</span>
        </a>
      </nav>

      <main className="docs-content">
        <header className="docs-header">
          <BookOpen size={48} className="docs-icon" />
          <span className="docs-kicker">Marketing to Technical Handoff</span>
          <h1>Documentation Gateway</h1>
          <p className="docs-tagline">
            Aeon Foundation explains why the stack matters. The docs explain how
            to build it, operate it, and go deeper through the book library.
          </p>
        </header>

        <section className="docs-section">
          <div className="section-heading">
            <h2>Start With The Right Depth</h2>
            <p>
              Choose whether you need orientation, reference, or a long-form
              deep dive.
            </p>
          </div>
          <div className="launch-grid">
            {launchPads.map((link) => (
              <a key={link.title} href={link.href} className="launch-card">
                <div className="launch-icon">{link.icon}</div>
                <div>
                  <h3>{link.title}</h3>
                  <p>{link.description}</p>
                </div>
              </a>
            ))}
          </div>
        </section>

        <section className="docs-section">
          <div className="section-heading">
            <h2>Follow The Gnosis Procession</h2>
            <p>
              For Gnosis specifically, the correct order is not “docs first.”
              Taste the language in Church, then deepen through docs and Book
              153, then move into IDE, Forge, and AeonOS when the evaluation
              turns into real work.
            </p>
          </div>
          <div className="launch-grid">
            {gnosisProcession.map((step) => (
              <a key={step.title} href={step.href} className="launch-card">
                <div className="launch-icon">
                  <BookOpen size={24} />
                </div>
                <div>
                  <p className="gnosis-procession-label">{step.label}</p>
                  <h3>{step.title}</h3>
                  <p>{step.description}</p>
                </div>
              </a>
            ))}
          </div>
        </section>

        <section className="docs-section">
          <div className="section-heading">
            <h2>Follow A Track</h2>
            <p>
              Each track hands off to typed docs first, then the right API and
              the first book in the series.
            </p>
          </div>
          <div className="track-grid">
            {tracks.map((track) => {
              const apiHref = getFoundationApiTrackUrl(track.id);
              const bookHref = getFoundationTrackBookUrl(track.id);
              return (
                <article key={track.id} className="track-card">
                  <div className="track-head">
                    <div className="track-icon">
                      {TRACK_ICONS[track.id] ?? <BookOpen size={24} />}
                    </div>
                    <div>
                      <h3>{track.label}</h3>
                      <p>{track.summary}</p>
                    </div>
                  </div>

                  <div className="track-meta">
                    <span>{track.ebookSlugs.length} curated books</span>
                    <span>
                      {apiHref
                        ? 'typed docs + api + books'
                        : 'typed docs + books'}
                    </span>
                  </div>

                  <div className="track-actions">
                    <a
                      href={getFoundationDocsTrackUrl(track.id)}
                      className="track-link primary"
                    >
                      Open Docs
                    </a>
                    {apiHref && (
                      <a href={apiHref} className="track-link">
                        API
                      </a>
                    )}
                    {bookHref && (
                      <a href={bookHref} className="track-link">
                        Read A Book
                      </a>
                    )}
                  </div>
                </article>
              );
            })}
          </div>
        </section>
      </main>

      <style>{`
        .docs-page {
          min-height: 100vh;
          background:
            radial-gradient(circle at top, rgba(34, 211, 238, 0.12), transparent 28%),
            linear-gradient(180deg, rgba(15, 23, 42, 0.98), rgba(2, 6, 23, 1));
          color: var(--color-text);
        }

        .docs-nav {
          padding: var(--space-lg) var(--space-xl);
          border-bottom: 1px solid rgba(148, 163, 184, 0.14);
        }

        .back-link {
          display: inline-flex;
          align-items: center;
          gap: var(--space-sm);
          color: var(--color-text-secondary);
          text-decoration: none;
          font-size: 0.9rem;
          transition: color 0.2s ease;
        }

        .back-link:hover {
          color: var(--color-text);
        }

        .docs-content {
          max-width: 1100px;
          margin: 0 auto;
          padding: var(--space-3xl) var(--space-xl) calc(var(--space-3xl) * 1.4);
        }

        .docs-header {
          margin-bottom: var(--space-3xl);
        }

        .docs-kicker {
          display: inline-flex;
          margin-bottom: var(--space-md);
          padding: 0.4rem 0.75rem;
          border-radius: 999px;
          font-size: 0.72rem;
          font-weight: 700;
          letter-spacing: 0.14em;
          text-transform: uppercase;
          color: #67e8f9;
          border: 1px solid rgba(103, 232, 249, 0.22);
          background: rgba(8, 47, 73, 0.28);
        }

        .docs-icon {
          color: var(--color-accent);
          margin-bottom: var(--space-md);
        }

        .docs-header h1 {
          margin-bottom: var(--space-sm);
          font-size: clamp(2.2rem, 5vw, 4rem);
        }

        .docs-tagline {
          max-width: 760px;
          font-size: 1.1rem;
          line-height: 1.7;
          color: var(--color-text-secondary);
        }

        .docs-section + .docs-section {
          margin-top: var(--space-3xl);
        }

        .gnosis-procession-label {
          margin: 0 0 var(--space-xs);
          color: #67e8f9;
          font-size: 0.72rem;
          font-weight: 700;
          letter-spacing: 0.12em;
          text-transform: uppercase;
        }

        .section-heading {
          margin-bottom: var(--space-xl);
        }

        .section-heading p {
          max-width: 760px;
          color: var(--color-text-secondary);
        }

        .launch-grid,
        .track-grid {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(240px, 1fr));
          gap: var(--space-lg);
        }

        .launch-card,
        .track-card {
          border-radius: 20px;
          border: 1px solid rgba(148, 163, 184, 0.14);
          background:
            linear-gradient(180deg, rgba(15, 23, 42, 0.88), rgba(2, 6, 23, 0.96));
          box-shadow:
            inset 0 1px 0 rgba(255, 255, 255, 0.04),
            0 24px 48px rgba(2, 6, 23, 0.22);
        }

        .launch-card {
          display: flex;
          gap: 0.9rem;
          align-items: flex-start;
          padding: 1.1rem;
          color: inherit;
          text-decoration: none;
        }

        .launch-card:hover,
        .track-card:hover {
          border-color: rgba(34, 211, 238, 0.28);
          transform: translateY(-2px);
        }

        .launch-icon,
        .track-icon {
          display: inline-flex;
          align-items: center;
          justify-content: center;
          width: 48px;
          height: 48px;
          border-radius: 14px;
          background: rgba(8, 47, 73, 0.32);
          color: #67e8f9;
          flex-shrink: 0;
        }

        .launch-card h3,
        .track-card h3 {
          margin: 0 0 0.35rem 0;
        }

        .launch-card p,
        .track-card p {
          margin: 0;
          color: var(--color-text-secondary);
          line-height: 1.6;
        }

        .track-card {
          padding: 1.25rem;
        }

        .track-head {
          display: flex;
          gap: 0.9rem;
          align-items: flex-start;
        }

        .track-meta {
          display: flex;
          flex-wrap: wrap;
          gap: 0.55rem;
          margin-top: 1rem;
          font-size: 0.78rem;
          letter-spacing: 0.04em;
          color: #7dd3fc;
        }

        .track-meta span {
          padding: 0.28rem 0.55rem;
          border-radius: 999px;
          border: 1px solid rgba(56, 189, 248, 0.18);
          background: rgba(8, 47, 73, 0.28);
        }

        .track-actions {
          display: flex;
          flex-wrap: wrap;
          gap: 0.75rem;
          margin-top: 1.15rem;
        }

        .track-link {
          display: inline-flex;
          align-items: center;
          justify-content: center;
          min-height: 40px;
          padding: 0.65rem 0.95rem;
          border-radius: 12px;
          border: 1px solid rgba(148, 163, 184, 0.18);
          color: inherit;
          text-decoration: none;
          font-weight: 600;
          background: rgba(15, 23, 42, 0.55);
        }

        .track-link.primary {
          border-color: rgba(34, 211, 238, 0.3);
          background: linear-gradient(135deg, rgba(13, 148, 136, 0.22), rgba(8, 47, 73, 0.45));
        }

        @media (max-width: 720px) {
          .docs-content {
            padding-inline: 1rem;
          }

          .track-actions {
            flex-direction: column;
          }

          .track-link {
            width: 100%;
          }
        }
      `}</style>
    </div>
  );
}
