// @a0n/.ignored_gnosis/forkjoin-app/client/pages/EcosystemProductPage.tsx
import {
  Activity,
  ArrowLeft,
  ArrowRight,
  BookOpen,
  Brain,
  Command,
  Compass,
  Database,
  FileSearch,
  Github,
  GitBranch,
  Globe,
  Hammer,
  HardDrive,
  KeyRound,
  Lock,
  MessageSquare,
  Monitor,
  Orbit,
  Radio,
  RefreshCw,
  Settings,
  Shield,
  Sparkles,
  Wallet,
  Waves,
  Workflow,
  Zap,
  type LucideIcon,
} from '@a0n/aeon-icons';
import {
  getFoundationProduct,
  getFoundationStackNeighbors,
  type FoundationProductId,
} from '../components/stack-model';

const productIconMap: Record<string, LucideIcon> = {
  activity: Activity,
  command: Command,
  compass: Compass,
  database: Database,
  'file-search': FileSearch,
  'git-branch': GitBranch,
  globe: Globe,
  hammer: Hammer,
  'hard-drive': HardDrive,
  'key-round': KeyRound,
  lock: Lock,
  'message-square': MessageSquare,
  monitor: Monitor,
  orbit: Orbit,
  radio: Radio,
  'refresh-cw': RefreshCw,
  settings: Settings,
  shield: Shield,
  sparkles: Sparkles,
  wallet: Wallet,
  waves: Waves,
  workflow: Workflow,
  brain: Brain,
  zap: Zap,
};

interface EcosystemProductPageProps {
  productId: FoundationProductId;
}

/**
 * Renders the Ecosystem Product Page view.
 */
export function EcosystemProductPage({ productId }: EcosystemProductPageProps) {
  const product = getFoundationProduct(productId);
  const ProductIcon = productIconMap[product.icon];
  const { above, below } = getFoundationStackNeighbors(productId);
  const relatedProductIds =
    (product.relatedProductIds as readonly FoundationProductId[] | undefined) ??
    [];
  const relatedProducts = relatedProductIds.map((relatedId) =>
    getFoundationProduct(relatedId)
  );

  return (
    <div
      className="product-page ecosystem-product-page"
      style={
        {
          '--product-color': product.color,
        } as React.CSSProperties
      }
    >
      <nav className="product-nav">
        <a href="/" className="back-link">
          <ArrowLeft size={20} />
          <span>Back to Aeon Foundation</span>
        </a>
      </nav>

      <main className="product-content">
        <header className="product-header">
          <ProductIcon size={64} className="product-icon" />
          <span className="product-badge">{product.layer}</span>
          <h1>{product.name}</h1>
          <p className="product-tagline">{product.tagline}</p>
          <div className="product-actions">
            <a href="/" className="button button-primary">
              <ArrowRight size={18} />
              Back to stack
            </a>
            {product.docsUrl ? (
              <a
                href={product.docsUrl}
                className="button button-secondary"
                target="_blank"
                rel="noopener noreferrer"
              >
                <BookOpen size={18} />
                Documentation
              </a>
            ) : null}
            {product.sourceUrl ? (
              <a
                href={product.sourceUrl}
                className="button button-secondary"
                target="_blank"
                rel="noopener noreferrer"
              >
                <Github size={18} />
                Source
              </a>
            ) : null}
          </div>
        </header>

        <section className="product-section">
          <h2>Role in the Stack</h2>
          <div className="product-essay">
            <p>{product.description}</p>
            <p>
              In this model, <strong>{product.name}</strong> sits in the{' '}
              <strong>{product.layer}</strong> layer and connects the
              experiences people touch to the sync, storage, and trust layers
              underneath.
            </p>
          </div>
          {product.install ? (
            <pre className="install-code">{product.install}</pre>
          ) : null}
        </section>

        <section className="product-section">
          <h2>Core Capabilities</h2>
          <div className="capability-grid">
            {product.highlights.map((highlight) => (
              <div key={highlight} className="capability-card">
                <span className="capability-dot" aria-hidden="true" />
                <p>{highlight}</p>
              </div>
            ))}
          </div>
        </section>

        <section className="product-section">
          <h2>Stack Neighbors</h2>
          <div className="neighbor-grid">
            {above ? (
              <a href={above.href} className="neighbor-card">
                <span className="neighbor-label">Above</span>
                <strong>{above.name}</strong>
                <span>{above.layer}</span>
              </a>
            ) : null}
            {below ? (
              <a href={below.href} className="neighbor-card">
                <span className="neighbor-label">Below</span>
                <strong>{below.name}</strong>
                <span>{below.layer}</span>
              </a>
            ) : null}
          </div>
          {relatedProducts.length > 0 ? (
            <div className="related-product-cloud">
              {relatedProducts.map((relatedProduct) => (
                <a
                  key={relatedProduct.id}
                  href={relatedProduct.href}
                  className="related-product-chip"
                >
                  {relatedProduct.name}
                </a>
              ))}
            </div>
          ) : null}
        </section>

        {(product.docsUrl || product.sourceUrl) && (
          <section className="product-section">
            <h2>Launch Points</h2>
            <div className="resource-grid">
              {product.docsUrl ? (
                <a
                  href={product.docsUrl}
                  className="resource-card"
                  target="_blank"
                  rel="noopener noreferrer"
                >
                  <BookOpen size={24} />
                  <h3>Documentation</h3>
                  <p>Reference material, package docs, and supporting books.</p>
                </a>
              ) : null}
              {product.sourceUrl ? (
                <a
                  href={product.sourceUrl}
                  className="resource-card"
                  target="_blank"
                  rel="noopener noreferrer"
                >
                  <Github size={24} />
                  <h3>Source</h3>
                  <p>
                    Repository entry point for the underlying implementation.
                  </p>
                </a>
              ) : null}
            </div>
          </section>
        )}
      </main>

      <style>{`
        .ecosystem-product-page {
          min-height: 100vh;
          background:
            radial-gradient(circle at top, color-mix(in srgb, var(--product-color) 18%, transparent) 0%, transparent 36%),
            var(--color-bg);
        }

        .product-nav {
          padding: var(--space-lg) var(--space-xl);
          border-bottom: 1px solid var(--color-bg-tertiary);
        }

        .back-link {
          display: inline-flex;
          align-items: center;
          gap: var(--space-sm);
          color: var(--color-text-secondary);
          text-decoration: none;
          transition: color 0.2s ease;
        }

        .back-link:hover {
          color: var(--color-text);
        }

        .product-content {
          max-width: 920px;
          margin: 0 auto;
          padding: var(--space-3xl) var(--space-xl);
        }

        .product-header {
          text-align: center;
          margin-bottom: var(--space-3xl);
        }

        .product-icon {
          display: block;
          margin: 0 auto var(--space-lg);
          color: var(--product-color);
        }

        .product-badge {
          display: inline-block;
          padding: 0.25rem 0.75rem;
          margin-bottom: var(--space-md);
          border-radius: 999px;
          border: 1px solid var(--product-color);
          color: var(--product-color);
          font-size: 0.78rem;
          letter-spacing: 0.12em;
          text-transform: uppercase;
        }

        .product-header h1 {
          margin-bottom: var(--space-md);
          font-size: clamp(2.6rem, 7vw, 4rem);
        }

        .product-tagline {
          max-width: 680px;
          margin: 0 auto var(--space-xl);
          color: var(--color-text-secondary);
          font-size: 1.2rem;
          line-height: 1.7;
        }

        .product-actions {
          display: flex;
          gap: var(--space-sm);
          justify-content: center;
          flex-wrap: wrap;
        }

        .product-actions .button {
          display: inline-flex;
          align-items: center;
          gap: var(--space-xs);
        }

        .product-section {
          margin-bottom: var(--space-3xl);
        }

        .product-section h2 {
          margin-bottom: var(--space-lg);
          padding-bottom: var(--space-sm);
          border-bottom: 1px solid var(--color-bg-tertiary);
        }

        .product-essay {
          display: grid;
          gap: var(--space-md);
          margin-bottom: var(--space-lg);
        }

        .product-essay p {
          margin: 0;
          color: var(--color-text-secondary);
          line-height: 1.85;
        }

        .install-code {
          margin: 0;
          padding: var(--space-lg) var(--space-xl);
          border-radius: 16px;
          border: 1px solid color-mix(in srgb, var(--product-color) 32%, rgba(255, 255, 255, 0.08));
          background: rgba(0, 0, 0, 0.2);
          color: var(--product-color);
          font-family: var(--font-mono);
          overflow-x: auto;
        }

        .capability-grid {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(220px, 1fr));
          gap: var(--space-md);
        }

        .capability-card,
        .neighbor-card,
        .resource-card {
          background: rgba(255, 255, 255, 0.03);
          border: 1px solid rgba(255, 255, 255, 0.08);
          border-radius: 18px;
          padding: var(--space-lg);
        }

        .capability-card {
          display: flex;
          gap: var(--space-sm);
          align-items: flex-start;
        }

        .capability-card p {
          margin: 0;
          color: var(--color-text-secondary);
          line-height: 1.6;
        }

        .capability-dot {
          width: 10px;
          height: 10px;
          margin-top: 0.4rem;
          border-radius: 999px;
          background: var(--product-color);
          flex-shrink: 0;
        }

        .neighbor-grid,
        .resource-grid {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(220px, 1fr));
          gap: var(--space-md);
        }

        .neighbor-card,
        .resource-card {
          text-decoration: none;
          color: inherit;
        }

        .neighbor-card strong,
        .resource-card h3 {
          display: block;
          margin-bottom: 0.35rem;
        }

        .neighbor-card span,
        .resource-card p {
          color: var(--color-text-secondary);
        }

        .neighbor-label {
          display: inline-block;
          margin-bottom: 0.65rem;
          color: var(--product-color);
          font-size: 0.78rem;
          letter-spacing: 0.12em;
          text-transform: uppercase;
        }

        .related-product-cloud {
          display: flex;
          flex-wrap: wrap;
          gap: 10px;
          margin-top: var(--space-lg);
        }

        .related-product-chip {
          padding: 0.55rem 0.85rem;
          border-radius: 999px;
          text-decoration: none;
          color: var(--color-text);
          border: 1px solid color-mix(in srgb, var(--product-color) 42%, rgba(255, 255, 255, 0.08));
          background: color-mix(in srgb, var(--product-color) 14%, rgba(255, 255, 255, 0.03));
        }
      `}</style>
    </div>
  );
}
