// @a0n/.ignored_gnosis/forkjoin-app/client/pages/FluxPage.tsx
/**
 * FluxPage - Dedicated page for Aeon-Flux framework
 *
 * Edge-first rendering framework with ESI and collaborative features.
 */

import {
  ArrowLeft,
  Waves,
  Globe,
  Brain,
  GitBranch,
  Zap,
  Users,
  Languages,
  Github,
  BookOpen,
  Code,
} from '@a0n/aeon-icons';
import { ArchitectureVisualization } from '../components/Flux3D/ArchitectureVisualization';
import { getFoundationDocsTrackUrl } from '../docsLinks';

/**
 * Renders the Flux Page view.
 */
export function FluxPage() {
  return (
    <div className="product-page flux-page">
      <nav className="product-nav">
        <a href="/" className="back-link">
          <ArrowLeft size={20} />
          <span>Back to Aeon Foundation</span>
        </a>
      </nav>

      <main className="product-content">
        <header className="product-header">
          <Waves size={64} className="product-icon" />
          <span className="product-badge">The Page Framework</span>
          <h1>Aeon-Flux</h1>
          <p className="product-tagline">
            Edge-first rendering. AI at render time. Instant navigation.
            Everything you wanted from that other framework, but couldn't have.
          </p>
          <div className="product-actions">
            <a
              href="https://github.com/forkjoin-ai/aeon-flux"
              className="button button-primary"
              target="_blank"
              rel="noopener noreferrer"
            >
              <Github size={18} />
              View on GitHub
            </a>
            <a
              href={getFoundationDocsTrackUrl('aeon-flux')}
              className="button button-secondary"
              target="_blank"
              rel="noopener noreferrer"
            >
              <BookOpen size={18} />
              Documentation
            </a>
          </div>
        </header>

        <section className="product-section">
          <h2>Installation</h2>
          <pre className="install-code">npx create-aeon-flux my-app</pre>
        </section>

        <section className="product-section">
          <h2>Performance</h2>
          <div className="metrics-row">
            <div className="metric-card">
              <span className="metric-value">180KB</span>
              <span className="metric-label">Runtime size</span>
              <span className="metric-comparison">vs 500KB+</span>
            </div>
            <div className="metric-card">
              <span className="metric-value">7ms</span>
              <span className="metric-label">Edge render</span>
              <span className="metric-comparison">vs 50ms+</span>
            </div>
            <div className="metric-card">
              <span className="metric-value">0ms</span>
              <span className="metric-label">Navigation</span>
              <span className="metric-comparison">Speculative</span>
            </div>
          </div>
        </section>

        <section className="product-section">
          <h2>The Manifesto</h2>
          <div className="manifesto-block">
            <h3>
              The CMS <em>is</em> the website.
            </h3>
            <p>
              Edit visually in the browser. Changes generate TSX. A PR is
              created. CI deploys. No developer context-switching. No separate
              admin panel. The website and the CMS are one.
            </p>
          </div>
        </section>

        <section className="product-section">
          <h2>Core Features</h2>
          <div className="features-grid">
            <FeatureCard
              icon={<Globe size={32} />}
              title="Edge-First Rendering"
              description="Every page renders in WASM at the edge. No origin server latency. Instant from anywhere on Earth."
            />
            <FeatureCard
              icon={<Brain size={32} />}
              title="Edge Side Inference (ESI)"
              description="AI at render time. Personalize content with LLM inference before the page reaches the browser."
            />
            <FeatureCard
              icon={<GitBranch size={32} />}
              title="Hyperpersonalized Routing"
              description="Routes adapt to the user. Time, location, emotion state, tier. Every visitor gets their own experience."
            />
            <FeatureCard
              icon={<Zap size={32} />}
              title="Speculative Pre-rendering"
              description="Pages render before you click. Hover prediction. Click = instant. Zero network delay."
            />
            <FeatureCard
              icon={<Users size={32} />}
              title="Real-time Collaboration"
              description="Multiple editors, same page, same time. CRDT-based. Cursor tracking. No conflicts."
            />
            <FeatureCard
              icon={<Languages size={32} />}
              title="AI Translation"
              description="Automatic translation at the edge. ESI.Translate renders content in the user's language. 33 languages supported."
            />
          </div>
        </section>

        <section className="product-section">
          <h2>Quick Start</h2>
          <div className="code-example">
            <div className="code-header">
              <Code size={18} />
              <span>Basic Page</span>
            </div>
            <pre className="code-block">{`'use aeon';

import { useAeonPage, ESI } from '@a0n/aeon-flux';

export default function Dashboard() {
  const { presence, data, setData } = useAeonPage();

  return (
    <div>
      {/* AI-powered personalized greeting */}
      <ESI.Infer prompt="Greet the user based on time and mood">
        {(greeting) => <h1>{greeting}</h1>}
      </ESI.Infer>

      {/* See who else is here */}
      <p>{presence.count} people viewing this page</p>

      {/* Collaborative content - synced in real-time */}
      <textarea
        value={data.notes}
        onChange={(e) => setData({ notes: e.target.value })}
      />
    </div>
  );
}`}</pre>
          </div>
        </section>

        <section className="product-section">
          <h2>Advanced Usage</h2>
          <div className="code-example">
            <div className="code-header">
              <Code size={18} />
              <span>Edge Side Inference (ESI)</span>
            </div>
            <pre className="code-block">{`import { ESI } from '@a0n/aeon-flux';

export default function ProductPage({ product }) {
  return (
    <div>
      <h1>{product.name}</h1>

      {/* Generate personalized product description */}
      <ESI.Infer
        prompt={\`Write a compelling description for \${product.name}
                 that appeals to the user's interests\`}
        context={{ userPreferences: true }}
      >
        {(description) => <p>{description}</p>}
      </ESI.Infer>

      {/* Personalized recommendations */}
      <ESI.Recommend
        based="browsing-history"
        limit={4}
      >
        {(recommendations) => (
          <div className="grid">
            {recommendations.map(item => (
              <ProductCard key={item.id} {...item} />
            ))}
          </div>
        )}
      </ESI.Recommend>

      {/* Automatic translation to user's language */}
      <ESI.Translate>
        Add to cart - Free shipping available
      </ESI.Translate>
    </div>
  );
}`}</pre>
          </div>

          <div className="code-example">
            <div className="code-header">
              <Code size={18} />
              <span>Visual Editing</span>
            </div>
            <pre className="code-block">{`import { Editable, EditableText, EditableImage } from '@a0n/aeon-flux/edit';

export default function LandingPage() {
  return (
    <Editable>
      <section className="hero">
        {/* Click to edit inline */}
        <EditableText
          field="hero.title"
          as="h1"
          placeholder="Enter headline..."
        />

        <EditableText
          field="hero.subtitle"
          as="p"
          placeholder="Enter subtitle..."
        />

        {/* Drag and drop image replacement */}
        <EditableImage
          field="hero.image"
          alt="Hero image"
          sizes="100vw"
        />

        {/* Changes auto-save and generate PR */}
      </section>
    </Editable>
  );
}`}</pre>
          </div>

          <div className="code-example">
            <div className="code-header">
              <Code size={18} />
              <span>Hyperpersonalized Routing</span>
            </div>
            <pre className="code-block">{`// aeon.config.ts
import { defineConfig } from '@a0n/aeon-flux';

export default defineConfig({
  routing: {
    personalization: {
      // Route based on user context
      factors: ['time', 'location', 'tier', 'emotion'],

      rules: [
        {
          match: { time: 'night', location: 'EU' },
          rewrite: '/dark-mode-landing',
        },
        {
          match: { tier: 'premium' },
          rewrite: '/premium-dashboard',
        },
        {
          match: { emotion: 'frustrated' },
          rewrite: '/help-center',
        },
      ],
    },
  },
});`}</pre>
          </div>
        </section>

        <section className="product-section">
          <h2>Architecture</h2>
          <div className="arch-visualization">
            <ArchitectureVisualization />
          </div>
        </section>

        <section className="product-section">
          <h2>Resources</h2>
          <div className="resources-grid">
            <a
              href="https://github.com/forkjoin-ai/aeon-flux"
              className="resource-card"
              target="_blank"
              rel="noopener noreferrer"
            >
              <Github size={24} />
              <h4>GitHub</h4>
              <p>Source code, issues, and discussions</p>
            </a>
            <a
              href={getFoundationDocsTrackUrl('aeon-flux')}
              className="resource-card"
              target="_blank"
              rel="noopener noreferrer"
            >
              <BookOpen size={24} />
              <h4>Documentation</h4>
              <p>Full API reference and guides</p>
            </a>
            <a
              href="https://github.com/forkjoin-ai/aeon-flux/tree/main/examples"
              className="resource-card"
              target="_blank"
              rel="noopener noreferrer"
            >
              <Code size={24} />
              <h4>Examples</h4>
              <p>Sample projects and templates</p>
            </a>
          </div>
        </section>
      </main>

      <style>{`
        .flux-page {
          --product-color: #10b981;
          --product-color-secondary: #14b8a6;
          min-height: 100vh;
          background: var(--color-bg);
        }

        .product-nav {
          padding: var(--space-lg) var(--space-xl);
          border-bottom: 1px solid var(--color-bg-tertiary);
        }

        .back-link {
          display: inline-flex;
          align-items: center;
          gap: var(--space-sm);
          color: var(--color-text-secondary);
          text-decoration: none;
          transition: color 0.2s ease;
        }

        .back-link:hover {
          color: var(--color-text);
        }

        .product-content {
          max-width: 900px;
          margin: 0 auto;
          padding: var(--space-3xl) var(--space-xl);
        }

        .product-header {
          text-align: center;
          margin-bottom: var(--space-3xl);
        }

        .product-icon {
          display: block;
          margin: 0 auto var(--space-lg);
          color: var(--product-color);
        }

        .product-badge {
          display: inline-block;
          padding: 0.25rem 0.75rem;
          font-size: 0.75rem;
          font-weight: 500;
          text-transform: uppercase;
          letter-spacing: 0.1em;
          color: var(--product-color);
          border: 1px solid var(--product-color);
          border-radius: var(--radius-full);
          margin-bottom: var(--space-md);
        }

        .product-header h1 {
          font-size: 3rem;
          margin-bottom: var(--space-md);
        }

        .product-tagline {
          font-size: 1.25rem;
          color: var(--color-text-secondary);
          max-width: 600px;
          margin: 0 auto var(--space-xl);
        }

        .product-actions {
          display: flex;
          gap: var(--space-md);
          justify-content: center;
          flex-wrap: wrap;
        }

        .product-actions .button {
          display: inline-flex;
          align-items: center;
          gap: var(--space-sm);
        }

        .product-section {
          margin-bottom: var(--space-3xl);
        }

        .product-section h2 {
          font-size: 1.5rem;
          margin-bottom: var(--space-lg);
          padding-bottom: var(--space-sm);
          border-bottom: 1px solid var(--color-bg-tertiary);
        }

        .install-code {
          font-family: var(--font-mono);
          font-size: 1.1rem;
          padding: var(--space-lg) var(--space-xl);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
          color: var(--product-color);
        }

        .metrics-row {
          display: flex;
          justify-content: center;
          gap: var(--space-xl);
          flex-wrap: wrap;
        }

        .metric-card {
          text-align: center;
          padding: var(--space-xl);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
          min-width: 160px;
        }

        .metric-value {
          display: block;
          font-size: 3rem;
          font-weight: 700;
          color: var(--product-color);
          line-height: 1;
        }

        .metric-label {
          display: block;
          font-size: 1rem;
          color: var(--color-text);
          margin-top: var(--space-sm);
        }

        .metric-comparison {
          display: block;
          font-size: 0.75rem;
          color: var(--color-text-muted);
          margin-top: var(--space-xs);
        }

        .manifesto-block {
          text-align: center;
          padding: var(--space-xl);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
        }

        .manifesto-block h3 {
          font-size: 1.75rem;
          margin-bottom: var(--space-md);
        }

        .manifesto-block em {
          color: var(--product-color);
          font-style: normal;
        }

        .manifesto-block p {
          max-width: 600px;
          margin: 0 auto;
          color: var(--color-text-secondary);
        }

        .features-grid {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(280px, 1fr));
          gap: var(--space-lg);
        }

        .feature-card {
          padding: var(--space-xl);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
        }

        .feature-card svg {
          color: var(--product-color-secondary);
          margin-bottom: var(--space-md);
        }

        .feature-card h4 {
          margin-bottom: var(--space-sm);
        }

        .feature-card p {
          color: var(--color-text-secondary);
          font-size: 0.9rem;
          margin: 0;
        }

        .code-example {
          margin-bottom: var(--space-xl);
        }

        .code-header {
          display: flex;
          align-items: center;
          gap: var(--space-sm);
          margin-bottom: var(--space-sm);
          color: var(--color-text-secondary);
          font-size: 0.9rem;
        }

        .code-header svg {
          color: var(--product-color);
        }

        .code-block {
          font-family: var(--font-mono);
          font-size: 0.85rem;
          line-height: 1.6;
          padding: var(--space-lg);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
          overflow-x: auto;
          margin: 0;
        }

        .arch-visualization {
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
          overflow: hidden;
          padding: var(--space-sm);
        }

        .resources-grid {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(250px, 1fr));
          gap: var(--space-lg);
        }

        .resource-card {
          padding: var(--space-xl);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
          text-decoration: none;
          color: inherit;
          transition: all 0.2s ease;
        }

        .resource-card:hover {
          border-color: var(--product-color);
          transform: translateY(-2px);
        }

        .resource-card svg {
          color: var(--product-color);
          margin-bottom: var(--space-md);
        }

        .resource-card h4 {
          margin-bottom: var(--space-xs);
        }

        .resource-card p {
          color: var(--color-text-secondary);
          font-size: 0.85rem;
          margin: 0;
        }
      `}</style>
    </div>
  );
}

interface FeatureCardProps {
  icon: React.ReactNode;
  title: string;
  description: string;
}

function FeatureCard({ icon, title, description }: FeatureCardProps) {
  return (
    <div className="feature-card">
      {icon}
      <h4>{title}</h4>
      <p>{description}</p>
    </div>
  );
}
