// @a0n/.ignored_gnosis/forkjoin-app/client/pages/NeuralPage.tsx
/**
 * NeuralPage - Dedicated page for Neural intelligence layer
 *
 * Privacy-preserving distributed learning with WebNN.
 */

import {
  ArrowLeft,
  Brain,
  Shield,
  Users,
  Cpu,
  Github,
  BookOpen,
  Code,
} from '@a0n/aeon-icons';
import { getFoundationDocsTrackUrl } from '../docsLinks';

/**
 * Renders the Neural Page view.
 */
export function NeuralPage() {
  return (
    <div className="product-page neural-page">
      <nav className="product-nav">
        <a href="/" className="back-link">
          <ArrowLeft size={20} />
          <span>Back to Aeon Foundation</span>
        </a>
      </nav>

      <main className="product-content">
        <header className="product-header">
          <Brain size={64} className="product-icon" />
          <span className="product-badge">Intelligence Layer</span>
          <h1>Neural</h1>
          <p className="product-tagline">
            WebNN-first inference with WebGPU fallback. Distributed learning
            that respects privacy.
          </p>
          <div className="product-actions">
            <a
              href="https://github.com/forkjoin-ai/gnosis"
              className="button button-primary"
              target="_blank"
              rel="noopener noreferrer"
            >
              <Github size={18} />
              View on GitHub
            </a>
            <a
              href={getFoundationDocsTrackUrl('neural')}
              className="button button-secondary"
              target="_blank"
              rel="noopener noreferrer"
            >
              <BookOpen size={18} />
              Documentation
            </a>
          </div>
        </header>

        <section className="product-section">
          <h2>Installation</h2>
          <pre className="install-code">npm install @a0n/gnosis</pre>
        </section>

        <section className="product-section">
          <h2>Core Concepts</h2>
          <div className="concepts-grid">
            <div className="concept-card">
              <div className="concept-visual">
                <div className="neuron-cluster">
                  <div className="neuron n1" />
                  <div className="neuron n2" />
                  <div className="neuron n3" />
                  <div className="neuron n4" />
                  <div className="neuron n5" />
                </div>
              </div>
              <h3>Federated Learning</h3>
              <p>
                Train models on user devices. Only gradients leave the edge.
                Your data never leaves your device, but the collective gets
                smarter.
              </p>
            </div>

            <div className="concept-card">
              <div className="concept-visual">
                <Shield size={48} className="concept-icon shield" />
              </div>
              <h3>Differential Privacy</h3>
              <p>
                Mathematical guarantees that protect individuals. Even with full
                access to outputs, your contribution remains hidden.
              </p>
            </div>

            <div className="concept-card">
              <div className="concept-visual">
                <Users size={48} className="concept-icon collective" />
              </div>
              <h3>Collective Intelligence</h3>
              <p>
                Every interaction makes the system smarter. Not by collecting
                your data — by learning from patterns that emerge.
              </p>
            </div>
          </div>
        </section>

        <section className="product-section">
          <h2>Core Principles</h2>
          <div className="principles-grid">
            <div className="principle">
              <span className="principle-num">01</span>
              <p>WebNN first — runs on any device with a neural accelerator.</p>
            </div>
            <div className="principle">
              <span className="principle-num">02</span>
              <p>Data stays on the edge. Models come to you.</p>
            </div>
            <div className="principle">
              <span className="principle-num">03</span>
              <p>Privacy is not a feature. It's the architecture.</p>
            </div>
            <div className="principle">
              <span className="principle-num">04</span>
              <p>Collective wisdom without collective surveillance.</p>
            </div>
          </div>
        </section>

        <section className="product-section">
          <h2>Privacy Guarantees</h2>
          <div className="privacy-metrics">
            <div className="privacy-metric">
              <span className="metric-value">10+</span>
              <span className="metric-label">
                Visitors needed for aggregation
              </span>
            </div>
            <div className="privacy-metric">
              <span className="metric-value">±0.1</span>
              <span className="metric-label">Noise added for privacy</span>
            </div>
            <div className="privacy-metric">
              <span className="metric-value">0</span>
              <span className="metric-label">
                Individual data points exposed
              </span>
            </div>
          </div>
        </section>

        <section className="product-section">
          <h2>Quick Start</h2>
          <div className="code-example">
            <div className="code-header">
              <Code size={18} />
              <span>WebNN Inference</span>
            </div>
            <pre className="code-block">{`import { Neural, WebNN } from '@a0n/gnosis';

// Initialize with WebNN (falls back to WebGPU)
const neural = new Neural({
  backend: 'webnn',    // 'webnn' | 'webgpu' | 'wasm'
  device: 'gpu',       // 'gpu' | 'npu' | 'cpu'
});

// Load a model
const model = await neural.load('emotion-classifier');

// Run inference on device
const result = await model.predict({
  text: 'I am feeling great today!',
});

console.log(result);
// { emotion: 'joy', confidence: 0.92 }`}</pre>
          </div>
        </section>

        <section className="product-section">
          <h2>Advanced Usage</h2>
          <div className="code-example">
            <div className="code-header">
              <Code size={18} />
              <span>Federated Learning</span>
            </div>
            <pre className="code-block">{`import { Neural, FederatedClient } from '@a0n/gnosis';

const neural = new Neural({ backend: 'webnn' });

// Join a federated learning session
const client = new FederatedClient({
  model: 'personalization-model',
  aggregator: 'wss://neural.aeonflux.dev/federated',
});

// Train locally on user data
await client.trainLocal({
  data: userInteractions,
  epochs: 3,
  batchSize: 32,
});

// Send only gradients (not data)
await client.submitGradients({
  // Differential privacy: add noise
  epsilon: 1.0,
  delta: 1e-5,

  // Secure aggregation: encrypt gradients
  encryption: 'threshold',
});

// Receive improved global model
const improvedModel = await client.receiveModel();`}</pre>
          </div>

          <div className="code-example">
            <div className="code-header">
              <Code size={18} />
              <span>Differential Privacy</span>
            </div>
            <pre className="code-block">{`import { Neural, PrivacyMechanism } from '@a0n/gnosis';

const neural = new Neural({ backend: 'webnn' });

// Configure privacy guarantees
const privacy = new PrivacyMechanism({
  epsilon: 1.0,        // Privacy budget
  delta: 1e-5,         // Failure probability
  mechanism: 'gaussian', // 'gaussian' | 'laplace'
});

// Aggregate with k-anonymity
const aggregate = await neural.aggregate({
  data: collectiveSignals,
  minK: 10,            // Require at least 10 contributors
  privacy,             // Apply differential privacy
});

// Individual contributions are mathematically hidden
console.log(aggregate);
// { mood: 0.72, energy: 0.65, focus: 0.58 }
// No way to identify individual inputs`}</pre>
          </div>

          <div className="code-example">
            <div className="code-header">
              <Code size={18} />
              <span>On-Device Model Fine-tuning</span>
            </div>
            <pre className="code-block">{`import { Neural, PersonalModel } from '@a0n/gnosis';

const neural = new Neural({ backend: 'webnn' });

// Create a personal model that stays on device
const personal = new PersonalModel({
  base: 'recommendation-v1',
  storage: 'indexeddb',  // Persists locally
});

// Fine-tune on user behavior
await personal.adapt({
  examples: userPreferences,
  learningRate: 0.001,
  preservePrivacy: true, // Never sends raw data
});

// Get personalized recommendations
const recommendations = await personal.recommend({
  context: currentContext,
  limit: 5,
});

// Optionally contribute to collective learning
if (userConsent) {
  await personal.contributeGradients({
    privacy: { epsilon: 0.5 },
  });
}`}</pre>
          </div>
        </section>

        <section className="product-section">
          <h2>Backend Support</h2>
          <div className="backend-grid">
            <div className="backend-card">
              <Cpu size={32} className="backend-icon webnn" />
              <h4>WebNN</h4>
              <p>
                Hardware-accelerated neural network inference. Uses NPU, GPU, or
                CPU as available.
              </p>
              <span className="backend-status recommended">Recommended</span>
            </div>
            <div className="backend-card">
              <Cpu size={32} className="backend-icon webgpu" />
              <h4>WebGPU</h4>
              <p>
                GPU compute shaders for neural network operations. Good fallback
                for WebNN.
              </p>
              <span className="backend-status fallback">Fallback</span>
            </div>
            <div className="backend-card">
              <Cpu size={32} className="backend-icon wasm" />
              <h4>WASM</h4>
              <p>
                Portable WebAssembly backend. Works everywhere, lower
                performance.
              </p>
              <span className="backend-status universal">Universal</span>
            </div>
          </div>
        </section>

        <section className="product-section">
          <h2>Resources</h2>
          <div className="resources-grid">
            <a
              href="https://github.com/forkjoin-ai/neural"
              className="resource-card"
              target="_blank"
              rel="noopener noreferrer"
            >
              <Github size={24} />
              <h4>GitHub</h4>
              <p>Source code, issues, and discussions</p>
            </a>
            <a
              href={getFoundationDocsTrackUrl('neural')}
              className="resource-card"
              target="_blank"
              rel="noopener noreferrer"
            >
              <BookOpen size={24} />
              <h4>Documentation</h4>
              <p>Full API reference and guides</p>
            </a>
            <a
              href="https://huggingface.co/affectively"
              className="resource-card"
              target="_blank"
              rel="noopener noreferrer"
            >
              <Brain size={24} />
              <h4>Models</h4>
              <p>Pre-trained models on HuggingFace</p>
            </a>
          </div>
        </section>
      </main>

      <style>{`
        .neural-page {
          --product-color: #336699;
          min-height: 100vh;
          background: linear-gradient(180deg, var(--color-bg) 0%, #0f0a1a 100%);
        }

        .product-nav {
          padding: var(--space-lg) var(--space-xl);
          border-bottom: 1px solid var(--color-bg-tertiary);
        }

        .back-link {
          display: inline-flex;
          align-items: center;
          gap: var(--space-sm);
          color: var(--color-text-secondary);
          text-decoration: none;
          transition: color 0.2s ease;
        }

        .back-link:hover {
          color: var(--color-text);
        }

        .product-content {
          max-width: 900px;
          margin: 0 auto;
          padding: var(--space-3xl) var(--space-xl);
        }

        .product-header {
          text-align: center;
          margin-bottom: var(--space-3xl);
        }

        .product-icon {
          display: block;
          margin: 0 auto var(--space-lg);
          color: var(--product-color);
        }

        .product-badge {
          display: inline-block;
          padding: 0.25rem 0.75rem;
          font-size: 0.75rem;
          font-weight: 500;
          text-transform: uppercase;
          letter-spacing: 0.1em;
          color: var(--product-color);
          border: 1px solid var(--product-color);
          border-radius: var(--radius-full);
          margin-bottom: var(--space-md);
        }

        .product-header h1 {
          font-size: 3rem;
          margin-bottom: var(--space-md);
        }

        .product-tagline {
          font-size: 1.25rem;
          color: var(--color-text-secondary);
          max-width: 600px;
          margin: 0 auto var(--space-xl);
        }

        .product-actions {
          display: flex;
          gap: var(--space-md);
          justify-content: center;
          flex-wrap: wrap;
        }

        .product-actions .button {
          display: inline-flex;
          align-items: center;
          gap: var(--space-sm);
        }

        .product-section {
          margin-bottom: var(--space-3xl);
        }

        .product-section h2 {
          font-size: 1.5rem;
          margin-bottom: var(--space-lg);
          padding-bottom: var(--space-sm);
          border-bottom: 1px solid var(--color-bg-tertiary);
        }

        .install-code {
          font-family: var(--font-mono);
          font-size: 1.1rem;
          padding: var(--space-lg) var(--space-xl);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
          color: var(--product-color);
        }

        .concepts-grid {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(280px, 1fr));
          gap: var(--space-xl);
        }

        .concept-card {
          text-align: center;
          padding: var(--space-xl);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 16px;
        }

        .concept-visual {
          height: 100px;
          display: flex;
          align-items: center;
          justify-content: center;
          margin-bottom: var(--space-lg);
        }

        .neuron-cluster {
          position: relative;
          width: 80px;
          height: 80px;
        }

        .neuron {
          position: absolute;
          width: 16px;
          height: 16px;
          background: radial-gradient(circle, var(--product-color) 0%, transparent 70%);
          border-radius: 50%;
          animation: pulse 2s ease-in-out infinite;
        }

        .n1 { top: 10%; left: 50%; transform: translateX(-50%); animation-delay: 0s; }
        .n2 { top: 35%; left: 15%; animation-delay: 0.3s; }
        .n3 { top: 35%; right: 15%; animation-delay: 0.6s; }
        .n4 { top: 60%; left: 30%; animation-delay: 0.9s; }
        .n5 { top: 60%; right: 30%; animation-delay: 1.2s; }

        @keyframes pulse {
          0%, 100% { opacity: 0.5; transform: scale(1); }
          50% { opacity: 1; transform: scale(1.3); }
        }

        .concept-icon {
          opacity: 0.8;
        }

        .concept-icon.shield {
          color: #336699;
        }

        .concept-icon.collective {
          color: #10b981;
        }

        .concept-card h3 {
          margin-bottom: var(--space-sm);
          color: var(--product-color);
        }

        .concept-card p {
          color: var(--color-text-secondary);
          font-size: 0.9rem;
          margin: 0;
        }

        .principles-grid {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(250px, 1fr));
          gap: var(--space-md);
        }

        .principle {
          display: flex;
          align-items: flex-start;
          gap: var(--space-sm);
          padding: var(--space-md);
          background: rgba(239, 68, 68, 0.05);
          border-left: 3px solid var(--product-color);
          border-radius: 0 8px 8px 0;
        }

        .principle-num {
          font-family: var(--font-mono);
          font-size: 0.8rem;
          color: var(--product-color);
          opacity: 0.7;
        }

        .principle p {
          margin: 0;
          font-size: 0.9rem;
          color: var(--color-text-secondary);
        }

        .privacy-metrics {
          display: flex;
          justify-content: center;
          gap: var(--space-3xl);
          padding: var(--space-xl);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 16px;
          flex-wrap: wrap;
        }

        .privacy-metric {
          text-align: center;
        }

        .metric-value {
          display: block;
          font-size: 2.5rem;
          font-weight: 700;
          color: var(--product-color);
        }

        .metric-label {
          font-size: 0.75rem;
          color: var(--color-text-muted);
          text-transform: uppercase;
          letter-spacing: 0.05em;
        }

        .backend-grid {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(250px, 1fr));
          gap: var(--space-lg);
        }

        .backend-card {
          padding: var(--space-xl);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
        }

        .backend-icon {
          margin-bottom: var(--space-md);
        }

        .backend-icon.webnn {
          color: #10b981;
        }

        .backend-icon.webgpu {
          color: #336699;
        }

        .backend-icon.wasm {
          color: #f59e0b;
        }

        .backend-card h4 {
          margin-bottom: var(--space-sm);
        }

        .backend-card p {
          color: var(--color-text-secondary);
          font-size: 0.85rem;
          margin-bottom: var(--space-md);
        }

        .backend-status {
          display: inline-block;
          font-size: 0.7rem;
          text-transform: uppercase;
          letter-spacing: 0.1em;
          padding: 4px 8px;
          border-radius: 4px;
        }

        .backend-status.recommended {
          background: rgba(16, 185, 129, 0.2);
          color: #10b981;
        }

        .backend-status.fallback {
          background: rgba(51, 102, 153, 0.2);
          color: #336699;
        }

        .backend-status.universal {
          background: rgba(245, 158, 11, 0.2);
          color: #f59e0b;
        }

        .code-example {
          margin-bottom: var(--space-xl);
        }

        .code-header {
          display: flex;
          align-items: center;
          gap: var(--space-sm);
          margin-bottom: var(--space-sm);
          color: var(--color-text-secondary);
          font-size: 0.9rem;
        }

        .code-header svg {
          color: var(--product-color);
        }

        .code-block {
          font-family: var(--font-mono);
          font-size: 0.85rem;
          line-height: 1.6;
          padding: var(--space-lg);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
          overflow-x: auto;
          margin: 0;
        }

        .resources-grid {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(250px, 1fr));
          gap: var(--space-lg);
        }

        .resource-card {
          padding: var(--space-xl);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
          text-decoration: none;
          color: inherit;
          transition: all 0.2s ease;
        }

        .resource-card:hover {
          border-color: var(--product-color);
          transform: translateY(-2px);
        }

        .resource-card svg {
          color: var(--product-color);
          margin-bottom: var(--space-md);
        }

        .resource-card h4 {
          margin-bottom: var(--space-xs);
        }

        .resource-card p {
          color: var(--color-text-secondary);
          font-size: 0.85rem;
          margin: 0;
        }
      `}</style>
    </div>
  );
}
