// @a0n/.ignored_gnosis/forkjoin-app/client/pages/Privacy.tsx
/**
 * Privacy Policy Page
 */

import { ArrowLeft, Shield, Eye, Lock, Database } from '@a0n/aeon-icons';

/**
 * Renders the Privacy view.
 */
export function Privacy() {
  return (
    <div className="legal-page">
      <nav className="legal-nav">
        <a href="/" className="back-link">
          <ArrowLeft size={20} />
          <span>Back to Aeon Foundation</span>
        </a>
      </nav>

      <main className="legal-content">
        <div className="legal-header">
          <Shield size={48} className="legal-icon" />
          <h1>Privacy Policy</h1>
          <p className="legal-date">Last updated: February 2026</p>
        </div>

        <section className="legal-section">
          <h2>Our Philosophy</h2>
          <p>
            Privacy isn't a feature we add—it's the architecture we build on.
            The Aeon Foundation is designed from first principles to respect
            your data sovereignty.
          </p>
          <div className="philosophy-grid">
            <div className="philosophy-item">
              <Lock size={24} />
              <h3>Local-First</h3>
              <p>
                Your data stays on your device unless you explicitly choose to
                sync.
              </p>
            </div>
            <div className="philosophy-item">
              <Eye size={24} />
              <h3>Zero-Knowledge</h3>
              <p>We can't see what you store. Even we don't have the keys.</p>
            </div>
            <div className="philosophy-item">
              <Database size={24} />
              <h3>Edge-Native</h3>
              <p>Data lives at the edge, not in centralized cloud databases.</p>
            </div>
          </div>
        </section>

        <section className="legal-section">
          <h2>What We Collect</h2>
          <h3>Anonymous Presence Data</h3>
          <p>
            When you visit aeonflux.dev, we create an anonymous session to
            enable real-time presence features. This includes:
          </p>
          <ul>
            <li>Anonymous session identifier (generated locally)</li>
            <li>Cursor position (if you choose to broadcast)</li>
            <li>Viewport dimensions (for responsive presence display)</li>
          </ul>
          <p>
            This data is <strong>never linked to your identity</strong> and is
            <strong> deleted immediately</strong> when you leave the page.
          </p>

          <h3>Aggregated Analytics</h3>
          <p>
            We collect aggregated, privacy-preserving analytics using
            k-anonymity and differential privacy. Individual behavior is never
            exposed—only collective patterns emerge.
          </p>
          <ul>
            <li>Minimum k=10 visitors before any aggregate is computed</li>
            <li>Random noise (ε=0.1) added to all aggregates</li>
            <li>No cookies, no fingerprinting, no tracking pixels</li>
          </ul>
        </section>

        <section className="legal-section">
          <h2>What We Don't Collect</h2>
          <ul>
            <li>Personal identifiers (name, email, phone)</li>
            <li>IP addresses (we don't log them)</li>
            <li>Browser fingerprints</li>
            <li>Third-party tracking cookies</li>
            <li>Cross-site identifiers</li>
            <li>Location data</li>
          </ul>
        </section>

        <section className="legal-section">
          <h2>Your Rights</h2>
          <p>
            Because we don't collect personal data, there's nothing to delete.
            But if you're using Aeon products that sync data, you have full
            control:
          </p>
          <ul>
            <li>
              <strong>Export</strong> all your data in standard formats
            </li>
            <li>
              <strong>Delete</strong> all synced data across devices
            </li>
            <li>
              <strong>Revoke</strong> any UCAN grants you've issued
            </li>
            <li>
              <strong>Disconnect</strong> any device from sync
            </li>
          </ul>
        </section>

        <section className="legal-section">
          <h2>Contact</h2>
          <p>
            Questions about privacy? Email us at{' '}
            <a href="mailto:privacy@aeonflux.dev">privacy@aeonflux.dev</a>
          </p>
        </section>
      </main>

      <style>{`
        .legal-page {
          min-height: 100vh;
          background: var(--color-bg);
          color: var(--color-text);
        }

        .legal-nav {
          padding: var(--space-lg) var(--space-xl);
          border-bottom: 1px solid var(--color-bg-tertiary);
        }

        .back-link {
          display: inline-flex;
          align-items: center;
          gap: var(--space-sm);
          color: var(--color-text-secondary);
          text-decoration: none;
          font-size: 0.9rem;
          transition: color 0.2s ease;
        }

        .back-link:hover {
          color: var(--color-text);
        }

        .legal-content {
          max-width: 800px;
          margin: 0 auto;
          padding: var(--space-3xl) var(--space-xl);
        }

        .legal-header {
          text-align: center;
          margin-bottom: var(--space-3xl);
        }

        .legal-icon {
          color: var(--color-accent);
          margin-bottom: var(--space-md);
        }

        .legal-header h1 {
          margin-bottom: var(--space-sm);
        }

        .legal-date {
          font-size: 0.875rem;
          color: var(--color-text-muted);
        }

        .legal-section {
          margin-bottom: var(--space-3xl);
        }

        .legal-section h2 {
          font-size: 1.5rem;
          margin-bottom: var(--space-lg);
          padding-bottom: var(--space-sm);
          border-bottom: 1px solid var(--color-bg-tertiary);
        }

        .legal-section h3 {
          font-size: 1.125rem;
          margin-top: var(--space-xl);
          margin-bottom: var(--space-md);
          color: var(--color-text-secondary);
        }

        .legal-section p {
          margin-bottom: var(--space-md);
          max-width: none;
        }

        .legal-section ul {
          list-style: none;
          padding: 0;
          margin: 0 0 var(--space-lg);
        }

        .legal-section li {
          padding: var(--space-sm) 0;
          padding-left: var(--space-lg);
          position: relative;
          color: var(--color-text-secondary);
        }

        .legal-section li::before {
          content: '—';
          position: absolute;
          left: 0;
          color: var(--color-accent);
        }

        .legal-section a {
          color: var(--color-accent);
        }

        .philosophy-grid {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(200px, 1fr));
          gap: var(--space-lg);
          margin-top: var(--space-xl);
        }

        .philosophy-item {
          padding: var(--space-lg);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
        }

        .philosophy-item svg {
          color: var(--color-accent);
          margin-bottom: var(--space-sm);
        }

        .philosophy-item h3 {
          margin: 0 0 var(--space-sm);
          font-size: 1rem;
          color: var(--color-text);
        }

        .philosophy-item p {
          margin: 0;
          font-size: 0.875rem;
        }
      `}</style>
    </div>
  );
}
