// @a0n/.ignored_gnosis/forkjoin-app/client/pages/RelayPage.tsx
/**
 * RelayPage - Dedicated page for Relay transport layer
 *
 * WebTransport mesh networking for real-time presence and sync.
 */

import { Suspense } from '@a0n/raect';
import {
  ArrowLeft,
  Network,
  Radio,
  Zap,
  Globe,
  Github,
  BookOpen,
  Code,
} from '@a0n/aeon-icons';
import { MeshVisualization } from '../components/Relay3D/MeshVisualization';
import { TopologyShowcase } from '../components/Relay3D/TopologyVisualization';
import { getFoundationDocsTrackUrl } from '../docsLinks';

/**
 * Renders the Relay Page view.
 */
export function RelayPage() {
  return (
    <div className="product-page relay-page">
      <nav className="product-nav">
        <a href="/" className="back-link">
          <ArrowLeft size={20} />
          <span>Back to Aeon Foundation</span>
        </a>
      </nav>

      <main className="product-content">
        <header className="product-header">
          <Network size={64} className="product-icon" />
          <span className="product-badge">Transport Layer</span>
          <h1>Relay</h1>
          <p className="product-tagline">
            WebTransport mesh networking for the real-time web. Because
            WebSockets were just the beginning.
          </p>
          <div className="product-actions">
            <a
              href="https://github.com/forkjoin-ai/dash"
              className="button button-primary"
              target="_blank"
              rel="noopener noreferrer"
            >
              <Github size={18} />
              View on GitHub
            </a>
            <a
              href={getFoundationDocsTrackUrl('dashrelay')}
              className="button button-secondary"
              target="_blank"
              rel="noopener noreferrer"
            >
              <BookOpen size={18} />
              Documentation
            </a>
          </div>
        </header>

        <section className="product-section">
          <h2>Installation</h2>
          <pre className="install-code">npm install @a0n/relay</pre>
        </section>

        <section className="product-section">
          <h2>Core Features</h2>
          <div className="features-grid">
            <div className="feature-card">
              <Zap size={40} className="feature-icon" />
              <h3>WebTransport Native</h3>
              <p>
                UDP-based transport with automatic fallback to WebSocket.
                Sub-10ms latency for real-time presence and collaboration.
              </p>
              <div className="latency-demo">
                <div
                  className="latency-bar"
                  style={{ '--latency': '8' } as React.CSSProperties}
                >
                  <span>Relay (WebTransport)</span>
                  <span>~8ms</span>
                </div>
                <div
                  className="latency-bar websocket"
                  style={{ '--latency': '45' } as React.CSSProperties}
                >
                  <span>WebSocket</span>
                  <span>~45ms</span>
                </div>
                <div
                  className="latency-bar polling"
                  style={{ '--latency': '150' } as React.CSSProperties}
                >
                  <span>HTTP Polling</span>
                  <span>~150ms</span>
                </div>
              </div>
            </div>

            <div className="feature-card mesh-card">
              <Network size={40} className="feature-icon" />
              <h3>Peer Mesh</h3>
              <p>
                Peer-to-peer mesh networking with intelligent relay routing. No
                central server bottleneck. Scales to millions.
              </p>
              <Suspense
                fallback={<div className="viz-loading">Loading...</div>}
              >
                <MeshVisualization />
              </Suspense>
            </div>

            <div className="feature-card">
              <Radio size={40} className="feature-icon" />
              <h3>Global Fanout</h3>
              <p>
                Messages propagate globally in under 50ms. Regional sharding
                with cross-region sync.
              </p>
              <div className="regions-list">
                <span className="region">us-east</span>
                <span className="region">us-west</span>
                <span className="region">eu-west</span>
                <span className="region">eu-central</span>
                <span className="region">asia-east</span>
                <span className="region more">+295 more</span>
              </div>
            </div>

            <div className="feature-card">
              <Globe size={40} className="feature-icon" />
              <h3>Edge-Native</h3>
              <p>
                Built on Cloudflare Durable Objects. WebSocket hibernation for
                cost efficiency. Automatic reconnection and offline queuing.
              </p>
              <div className="stats-row">
                <div className="stat">
                  <span className="stat-value">300+</span>
                  <span className="stat-label">PoPs</span>
                </div>
                <div className="stat">
                  <span className="stat-value">50ms</span>
                  <span className="stat-label">global</span>
                </div>
                <div className="stat">
                  <span className="stat-value">0</span>
                  <span className="stat-label">origin</span>
                </div>
              </div>
            </div>
          </div>
        </section>

        <section className="product-section">
          <h2>Network Topologies</h2>
          <p className="section-intro">
            Relay now rides on the Dash sync stack. Attach a Yjs document for
            binary message streams, a QDoc for topology-native CRDT sync, or an
            Automerge document for immutable collaborative state.
          </p>
          <Suspense
            fallback={
              <div className="viz-loading">Loading visualization...</div>
            }
          >
            <TopologyShowcase />
          </Suspense>
        </section>

        <section className="product-section">
          <h2>Quick Start</h2>
          <div className="code-example">
            <div className="code-header">
              <Code size={18} />
              <span>Basic Usage</span>
            </div>
            <pre className="code-block">{`import { Relay } from '@a0n/relay';
import { Doc } from 'yjs';

const doc = new Doc();
const encoder = new TextEncoder();
const decoder = new TextDecoder();

const relay = new Relay({
  roomName: 'presence:lobby',
  transport: 'websocket',
});

relay.on('connected', () => {
  console.log('Connected to relay');
});

relay.on('message', (peerId, payload) => {
  const message = JSON.parse(decoder.decode(payload));
  updateCursors(peerId, message);
});

await relay.connect(doc);

window.addEventListener('pointermove', (event) => {
  relay.broadcast(
    encoder.encode(JSON.stringify({ x: event.clientX, y: event.clientY }))
  );
});`}</pre>
          </div>
        </section>

        <section className="product-section">
          <h2>Advanced Usage</h2>
          <div className="code-example">
            <div className="code-header">
              <Code size={18} />
              <span>QDoc Sync</span>
            </div>
            <pre className="code-block">{`import { QDoc } from '@a0n/gnosis';
import { Relay } from '@a0n/relay';

const doc = new QDoc();
const notes = doc.getMap('notes');

const relay = new Relay({
  roomName: 'document:123',
  ucan: session.ucan,
});

relay.on('connected', () => {
  console.log('QDoc relay connected');
});

await relay.connectQDoc(doc);

notes.set('title', 'Shared outline');`}</pre>
          </div>

          <div className="code-example">
            <div className="code-header">
              <Code size={18} />
              <span>Automerge Documents</span>
            </div>
            <pre className="code-block">{`import * as Automerge from '@automerge/automerge';
import { Relay } from '@a0n/relay';

type CounterDoc = { count?: number };
let doc = Automerge.init<CounterDoc>();

const relay = new Relay({ roomName: 'counter:global' });

await relay.connectAutomerge(doc, {
  onChange: (nextDoc) => {
    doc = nextDoc;
    render(doc);
  },
});

doc = Automerge.change(doc, (draft) => {
  draft.count = (draft.count ?? 0) + 1;
});

relay.updateAutomergeDoc(doc);`}</pre>
          </div>

          <div className="code-example">
            <div className="code-header">
              <Code size={18} />
              <span>Connection Lifecycle</span>
            </div>
            <pre className="code-block">{`import { Relay } from '@a0n/relay';
import { Doc } from 'yjs';

const doc = new Doc();
const relay = new Relay({
  roomName: 'presence:ops',
  url: 'wss://relay.dashrelay.com/relay/sync',
});

relay.on('connected', () => showStatus('online'));
relay.on('disconnected', () => showStatus('offline'));
relay.on('peerJoined', (peerId) => console.log('Peer joined', peerId));
relay.on('peerLeft', (peerId) => console.log('Peer left', peerId));
relay.on('error', (error) => console.error(error));

await relay.connect(doc);

window.addEventListener('beforeunload', () => relay.disconnect());`}</pre>
          </div>
        </section>

        <section className="product-section">
          <h2>Resources</h2>
          <div className="resources-grid">
            <a
              href="https://github.com/forkjoin-ai/dash"
              className="resource-card"
              target="_blank"
              rel="noopener noreferrer"
            >
              <Github size={24} />
              <h4>GitHub</h4>
              <p>Source code, issues, and discussions</p>
            </a>
            <a
              href={getFoundationDocsTrackUrl('dashrelay')}
              className="resource-card"
              target="_blank"
              rel="noopener noreferrer"
            >
              <BookOpen size={24} />
              <h4>Documentation</h4>
              <p>Full API reference and guides</p>
            </a>
            <a
              href="https://github.com/forkjoin-ai/emotions/tree/main/packages/relay"
              className="resource-card"
              target="_blank"
              rel="noopener noreferrer"
            >
              <Code size={24} />
              <h4>Examples</h4>
              <p>Sample projects and use cases</p>
            </a>
          </div>
        </section>
      </main>

      <style>{`
        .relay-page {
          --product-color: #06b6d4;
          min-height: 100vh;
          background: var(--color-bg);
        }

        .product-nav {
          padding: var(--space-lg) var(--space-xl);
          border-bottom: 1px solid var(--color-bg-tertiary);
        }

        .back-link {
          display: inline-flex;
          align-items: center;
          gap: var(--space-sm);
          color: var(--color-text-secondary);
          text-decoration: none;
          transition: color 0.2s ease;
        }

        .back-link:hover {
          color: var(--color-text);
        }

        .product-content {
          max-width: 900px;
          margin: 0 auto;
          padding: var(--space-3xl) var(--space-xl);
        }

        .product-header {
          text-align: center;
          margin-bottom: var(--space-3xl);
        }

        .product-icon {
          display: block;
          margin: 0 auto var(--space-lg);
          color: var(--product-color);
        }

        .product-badge {
          display: inline-block;
          padding: 0.25rem 0.75rem;
          font-size: 0.75rem;
          font-weight: 500;
          text-transform: uppercase;
          letter-spacing: 0.1em;
          color: var(--product-color);
          border: 1px solid var(--product-color);
          border-radius: var(--radius-full);
          margin-bottom: var(--space-md);
        }

        .product-header h1 {
          font-size: 3rem;
          margin-bottom: var(--space-md);
        }

        .product-tagline {
          font-size: 1.25rem;
          color: var(--color-text-secondary);
          max-width: 600px;
          margin: 0 auto var(--space-xl);
        }

        .product-actions {
          display: flex;
          gap: var(--space-md);
          justify-content: center;
          flex-wrap: wrap;
        }

        .product-actions .button {
          display: inline-flex;
          align-items: center;
          gap: var(--space-sm);
        }

        .product-section {
          margin-bottom: var(--space-3xl);
        }

        .product-section h2 {
          font-size: 1.5rem;
          margin-bottom: var(--space-lg);
          padding-bottom: var(--space-sm);
          border-bottom: 1px solid var(--color-bg-tertiary);
        }

        .install-code {
          font-family: var(--font-mono);
          font-size: 1.1rem;
          padding: var(--space-lg) var(--space-xl);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
          color: var(--product-color);
        }

        .features-grid {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(280px, 1fr));
          gap: var(--space-xl);
        }

        .feature-card {
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 16px;
          padding: var(--space-xl);
        }

        .feature-icon {
          color: var(--product-color);
          margin-bottom: var(--space-md);
        }

        .feature-card h3 {
          margin-bottom: var(--space-sm);
          color: var(--product-color);
        }

        .feature-card p {
          color: var(--color-text-secondary);
          margin-bottom: var(--space-lg);
        }

        .latency-demo {
          display: flex;
          flex-direction: column;
          gap: var(--space-sm);
        }

        .latency-bar {
          display: flex;
          justify-content: space-between;
          align-items: center;
          padding: var(--space-sm) var(--space-md);
          background: linear-gradient(
            90deg,
            rgba(6, 182, 212, 0.3) 0%,
            rgba(6, 182, 212, 0.1) calc(var(--latency) * 0.6%),
            transparent calc(var(--latency) * 0.6%)
          );
          border-radius: 6px;
          font-size: 0.85rem;
        }

        .latency-bar.websocket {
          background: linear-gradient(
            90deg,
            rgba(51, 102, 153, 0.3) 0%,
            rgba(51, 102, 153, 0.1) calc(var(--latency) * 0.6%),
            transparent calc(var(--latency) * 0.6%)
          );
        }

        .latency-bar.polling {
          background: linear-gradient(
            90deg,
            rgba(239, 68, 68, 0.3) 0%,
            rgba(239, 68, 68, 0.1) calc(var(--latency) * 0.6%),
            transparent calc(var(--latency) * 0.6%)
          );
        }

        .mesh-card {
          grid-column: span 2;
        }

        .viz-loading {
          height: 250px;
          display: flex;
          align-items: center;
          justify-content: center;
          color: var(--color-text-muted);
          font-size: 0.85rem;
        }

        .section-intro {
          color: var(--color-text-secondary);
          margin-bottom: var(--space-lg);
          font-size: 0.95rem;
        }

        @media (max-width: 768px) {
          .mesh-card {
            grid-column: span 1;
          }
        }

        .regions-list {
          display: flex;
          flex-wrap: wrap;
          gap: var(--space-xs);
        }

        .region {
          font-size: 0.7rem;
          padding: 4px 8px;
          background: rgba(6, 182, 212, 0.1);
          border: 1px solid rgba(6, 182, 212, 0.3);
          border-radius: 4px;
          color: var(--product-color);
        }

        .region.more {
          background: rgba(51, 102, 153, 0.1);
          border-color: rgba(51, 102, 153, 0.3);
          color: #336699;
        }

        .stats-row {
          display: flex;
          gap: var(--space-lg);
        }

        .stat {
          display: flex;
          flex-direction: column;
          align-items: center;
        }

        .stat-value {
          font-family: var(--font-mono);
          font-size: 1.5rem;
          font-weight: 600;
          color: var(--product-color);
        }

        .stat-label {
          font-size: 0.7rem;
          text-transform: uppercase;
          letter-spacing: 0.1em;
          color: var(--color-text-muted);
        }

        .code-example {
          margin-bottom: var(--space-xl);
        }

        .code-header {
          display: flex;
          align-items: center;
          gap: var(--space-sm);
          margin-bottom: var(--space-sm);
          color: var(--color-text-secondary);
          font-size: 0.9rem;
        }

        .code-header svg {
          color: var(--product-color);
        }

        .code-block {
          font-family: var(--font-mono);
          font-size: 0.85rem;
          line-height: 1.6;
          padding: var(--space-lg);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
          overflow-x: auto;
          margin: 0;
        }

        .resources-grid {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(250px, 1fr));
          gap: var(--space-lg);
        }

        .resource-card {
          padding: var(--space-xl);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
          text-decoration: none;
          color: inherit;
          transition: all 0.2s ease;
        }

        .resource-card:hover {
          border-color: var(--product-color);
          transform: translateY(-2px);
        }

        .resource-card svg {
          color: var(--product-color);
          margin-bottom: var(--space-md);
        }

        .resource-card h4 {
          margin-bottom: var(--space-xs);
        }

        .resource-card p {
          color: var(--color-text-secondary);
          font-size: 0.85rem;
          margin: 0;
        }
      `}</style>
    </div>
  );
}
