// @a0n/.ignored_gnosis/forkjoin-app/client/pages/Terms.tsx
/**
 * Terms of Service Page
 */

import {
  ArrowLeft,
  FileText,
  Scale,
  AlertTriangle,
  Check,
} from '@a0n/aeon-icons';

/**
 * Renders the Terms view.
 */
export function Terms() {
  return (
    <div className="legal-page">
      <nav className="legal-nav">
        <a href="/" className="back-link">
          <ArrowLeft size={20} />
          <span>Back to Aeon Foundation</span>
        </a>
      </nav>

      <main className="legal-content">
        <div className="legal-header">
          <FileText size={48} className="legal-icon" />
          <h1>Terms of Service</h1>
          <p className="legal-date">Last updated: February 2026</p>
        </div>

        <section className="legal-section">
          <h2>Plain English Summary</h2>
          <div className="summary-grid">
            <div className="summary-item allowed">
              <Check size={20} />
              <span>Use our open-source code freely under MIT/Apache-2.0</span>
            </div>
            <div className="summary-item allowed">
              <Check size={20} />
              <span>Build commercial products with Aeon Foundation</span>
            </div>
            <div className="summary-item allowed">
              <Check size={20} />
              <span>Modify and redistribute the source code</span>
            </div>
            <div className="summary-item warning">
              <AlertTriangle size={20} />
              <span>Don't use our trademarks without permission</span>
            </div>
            <div className="summary-item warning">
              <AlertTriangle size={20} />
              <span>No warranty—use at your own risk</span>
            </div>
          </div>
        </section>

        <section className="legal-section">
          <h2>1. Acceptance of Terms</h2>
          <p>
            By accessing or using aeonflux.dev or any Aeon Foundation products,
            you agree to be bound by these Terms of Service. If you disagree
            with any part, please don't use our services.
          </p>
        </section>

        <section className="legal-section">
          <h2>2. Open Source License</h2>
          <p>
            All Aeon Foundation source code is released under dual MIT and
            Apache-2.0 licenses. You may choose whichever license best suits
            your needs.
          </p>
          <p>This means you can:</p>
          <ul>
            <li>Use the code for any purpose, including commercial</li>
            <li>Modify, merge, publish, and distribute the code</li>
            <li>Sublicense and sell copies</li>
            <li>Use it in proprietary software</li>
          </ul>
          <p>
            The only requirement is to include the original copyright notice and
            license text.
          </p>
        </section>

        <section className="legal-section">
          <h2>3. Trademarks</h2>
          <p>
            "Aeon Foundation", "Aeon-Flux", "Aeon", "Dash", "Relay", "Aegis",
            "Neural", and "Edgework" are trademarks of Affectively, Inc.
          </p>
          <p>
            You may not use our trademarks in a way that suggests endorsement or
            affiliation without written permission. Fair use for attribution and
            reference is permitted.
          </p>
        </section>

        <section className="legal-section">
          <h2>4. Hosted Services</h2>
          <p>
            If you use our hosted services (aeonflux.dev, dashrelay.dev, etc.):
          </p>
          <ul>
            <li>You agree to use them lawfully and not for harm</li>
            <li>We may rate-limit or block abusive usage</li>
            <li>We provide no uptime guarantees for free tiers</li>
            <li>We may discontinue services with reasonable notice</li>
          </ul>
        </section>

        <section className="legal-section">
          <h2>5. Disclaimer of Warranties</h2>
          <p>
            THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
            EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
            MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
            NONINFRINGEMENT.
          </p>
        </section>

        <section className="legal-section">
          <h2>6. Limitation of Liability</h2>
          <p>
            IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
            CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
            TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
            SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
          </p>
        </section>

        <section className="legal-section">
          <h2>7. Changes to Terms</h2>
          <p>
            We may update these terms. Significant changes will be announced on
            our GitHub and Discord. Continued use after changes constitutes
            acceptance.
          </p>
        </section>

        <section className="legal-section">
          <h2>Contact</h2>
          <p>
            Questions? Email us at{' '}
            <a href="mailto:legal@aeonflux.dev">legal@aeonflux.dev</a>
          </p>
        </section>
      </main>

      <style>{`
        .legal-page {
          min-height: 100vh;
          background: var(--color-bg);
          color: var(--color-text);
        }

        .legal-nav {
          padding: var(--space-lg) var(--space-xl);
          border-bottom: 1px solid var(--color-bg-tertiary);
        }

        .back-link {
          display: inline-flex;
          align-items: center;
          gap: var(--space-sm);
          color: var(--color-text-secondary);
          text-decoration: none;
          font-size: 0.9rem;
          transition: color 0.2s ease;
        }

        .back-link:hover {
          color: var(--color-text);
        }

        .legal-content {
          max-width: 800px;
          margin: 0 auto;
          padding: var(--space-3xl) var(--space-xl);
        }

        .legal-header {
          text-align: center;
          margin-bottom: var(--space-3xl);
        }

        .legal-icon {
          color: var(--color-accent);
          margin-bottom: var(--space-md);
        }

        .legal-header h1 {
          margin-bottom: var(--space-sm);
        }

        .legal-date {
          font-size: 0.875rem;
          color: var(--color-text-muted);
        }

        .legal-section {
          margin-bottom: var(--space-3xl);
        }

        .legal-section h2 {
          font-size: 1.5rem;
          margin-bottom: var(--space-lg);
          padding-bottom: var(--space-sm);
          border-bottom: 1px solid var(--color-bg-tertiary);
        }

        .legal-section p {
          margin-bottom: var(--space-md);
          max-width: none;
        }

        .legal-section ul {
          list-style: none;
          padding: 0;
          margin: 0 0 var(--space-lg);
        }

        .legal-section li {
          padding: var(--space-sm) 0;
          padding-left: var(--space-lg);
          position: relative;
          color: var(--color-text-secondary);
        }

        .legal-section li::before {
          content: '—';
          position: absolute;
          left: 0;
          color: var(--color-accent);
        }

        .legal-section a {
          color: var(--color-accent);
        }

        .summary-grid {
          display: flex;
          flex-direction: column;
          gap: var(--space-sm);
        }

        .summary-item {
          display: flex;
          align-items: center;
          gap: var(--space-md);
          padding: var(--space-md) var(--space-lg);
          background: var(--color-bg-secondary);
          border-radius: 8px;
          font-size: 0.95rem;
        }

        .summary-item.allowed svg {
          color: #10b981;
        }

        .summary-item.warning svg {
          color: #f59e0b;
        }
      `}</style>
    </div>
  );
}
