// @a0n/gnosis/distributed-inference/scripts/preconditioning/fluency-gate.mjs
// fluency-gate.mjs — grammatical gate for the prompt-preconditioner.
//
// A preconditioner swaps a prompt's content words to true WordNet synonyms
// (table: ./synonyms.json, format {word: {n?:[...],v?:[...],a?:[...],r?:[...]}}
// where n=noun v=verb a=adjective r=adverb). Greedy swaps were sometimes
// UNGRAMMATICAL even when they preserved the model's answer:
//   "...difficult to complete" -> "...difficult to terminated" (broken inflection)
//   "...felt very happy"       -> "...felt real happy"          (adverb slot got an adjective)
//
// This gate filters a word's synonyms so the swap stays grammatical:
//   1. POS preservation  — candidate must come from the SAME POS bucket as the slot.
//   2. Inflection match  — if the surface word is plural / past / gerund / -ly,
//                          re-inflect the candidate to match; drop ones we can't.
//   3. Length / identity — drop candidates == the word or shorter than 3 chars.
//
// It composes downstream with answer-identity + compute gates: rejection here
// simply means "no swap" for that word (the caller falls back to the original).

import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";

const __dirname = dirname(fileURLToPath(import.meta.url));
const SYN_PATH = join(__dirname, "synonyms.json");

let _table = null;
function table() {
  if (!_table) _table = JSON.parse(readFileSync(SYN_PATH, "utf8"));
  return _table;
}

// ---------------------------------------------------------------------------
// English inflection (simple, deterministic).
// ---------------------------------------------------------------------------

const VOWELS = new Set(["a", "e", "i", "o", "u"]);
const isVowel = (c) => VOWELS.has(c);

// "Flat" adverbs: identical in adjective + adverb surface form, no -ly.
const FLAT_ADVERBS = new Set([
  "fast", "straight", "straightaway", "hard", "late", "early", "near",
  "high", "low", "deep", "wide", "right", "wrong",
]);

// base + plural (cats, boxes, parties)
/**
 * Handles the gnosis pluralize workflow.
 */
export function pluralize(base) {
  if (/(s|x|z|ch|sh)$/.test(base)) return base + "es";
  if (/[^aeiou]y$/.test(base)) return base.slice(0, -1) + "ies";
  return base + "s";
}

// base + 3rd-person-sing / verb-s — same rules as noun plural
/**
 * Handles the gnosis third Person workflow.
 */
export function thirdPerson(base) {
  return pluralize(base);
}

// base + past tense (-ed): liked, carried, stopped
/**
 * Handles the gnosis past Tense workflow.
 */
export function pastTense(base) {
  if (base.endsWith("e")) return base + "d";
  if (/[^aeiou]y$/.test(base)) return base.slice(0, -1) + "ied";
  if (needsDouble(base)) return base + base.slice(-1) + "ed";
  return base + "ed";
}

// base + gerund (-ing): liking, running, carrying
/**
 * Handles the gnosis gerund workflow.
 */
export function gerund(base) {
  if (base.endsWith("ee")) return base + "ing"; // seeing, freeing
  if (base.endsWith("e") && !base.endsWith("oe")) return base.slice(0, -1) + "ing";
  if (needsDouble(base)) return base + base.slice(-1) + "ing";
  return base + "ing";
}

// adjective -> adverb (-ly): quick->quickly, happy->happily, simple->simply, basic->basically.
// Returns null when the base can't be cleanly adverbized (already -ly handled by
// caller; -ing/-ed participial forms and a few irregulars don't take a plain -ly).
/**
 * Handles the gnosis adverbize workflow.
 */
export function adverbize(base) {
  if (base.endsWith("ly")) return base;               // already an adverb (quickly, promptly)
  if (base.endsWith("ing")) return null;              // flying->*flyingly: reject
  if (base.endsWith("ed")) return null;               // tired->*tiredly: reject (unreliable)
  if (FLAT_ADVERBS.has(base)) return base; // flat adverbs: already adverbial as-is
  if (/[^aeiou]y$/.test(base)) return base.slice(0, -1) + "ily"; // happy->happily
  if (base.endsWith("le") && !isVowel(base[base.length - 3])) return base.slice(0, -1) + "y"; // simple->simply
  if (base.endsWith("ic")) return base + "ally";      // basic->basically
  if (base.endsWith("ll")) return base + "y";         // full->fully (rare here)
  return base + "ly";
}

// Final-consonant doubling, conservatively: only for short, ~monosyllabic
// words ending in consonant-vowel-consonant where the preceding letter is NOT
// a vowel (so the final vowel is "short"). stop->stopped, run->running, but
// NOT nail (vowel-vowel-cons), function (two syllables), lead (vowel-vowel),
// prevail (vowel-vowel), lean (vowel-vowel).
const ROUGH_SYLLABLE = /[aeiouy]/g;
function needsDouble(base) {
  if (base.length < 3) return false;
  // Heuristic: doubling only on words that are roughly one syllable.
  const vowelGroups = base.match(/[aeiouy]+/g) || [];
  if (vowelGroups.length !== 1) return false;
  const a = base[base.length - 3];
  const b = base[base.length - 2];
  const c = base[base.length - 1];
  if (isVowel(c) || c === "w" || c === "x" || c === "y") return false; // never double w/x/y
  if (!isVowel(b)) return false;     // need a vowel before the final consonant
  if (isVowel(a)) return false;      // a vowel before that vowel = long vowel (lead, rain) -> no double
  return true;                       // C V C, single short vowel -> double
}

// ---------------------------------------------------------------------------
// Surface analysis: detect ending + recover the lemma to look up.
// ---------------------------------------------------------------------------

// Returns {ending, lemmas:[...]} where ending is one of:
//   "plural" | "past" | "gerund" | "adverb" | "base"
// and lemmas are candidate base forms to look up in the synonym table.
/**
 * Handles the gnosis analyze Surface workflow.
 */
export function analyzeSurface(surface) {
  const w = surface.toLowerCase();

  if (w.endsWith("ly") && w.length > 3) {
    const lemmas = [];
    lemmas.push(w.slice(0, -2)); // quickly->quick
    if (w.endsWith("ily")) lemmas.push(w.slice(0, -3) + "y"); // happily->happy
    if (w.endsWith("ally")) lemmas.push(w.slice(0, -4) + "ic"); // basically->basic
    if (w.endsWith("ly") && !w.endsWith("lly")) lemmas.push(w.slice(0, -1)); // simply->simpl(e)? handled below
    return { ending: "adverb", lemmas: dedupe(lemmas) };
  }

  if (w.endsWith("ing") && w.length > 4) {
    const stem = w.slice(0, -3);
    const lemmas = [stem, stem + "e"]; // run<-running(double handled), make<-making
    if (stem.length >= 2 && stem[stem.length - 1] === stem[stem.length - 2]) {
      lemmas.push(stem.slice(0, -1)); // running->run
    }
    return { ending: "gerund", lemmas: dedupe(lemmas) };
  }

  if (w.endsWith("ed") && w.length > 3) {
    const stem = w.slice(0, -2);
    const lemmas = [stem, w.slice(0, -1)]; // wanted->want, liked->like
    if (stem.endsWith("i")) lemmas.push(stem.slice(0, -1) + "y"); // carried->carry
    if (stem.length >= 2 && stem[stem.length - 1] === stem[stem.length - 2]) {
      lemmas.push(stem.slice(0, -1)); // stopped->stop
    }
    return { ending: "past", lemmas: dedupe(lemmas) };
  }

  if ((w.endsWith("es") && w.length > 4) || (w.endsWith("s") && w.length > 3)) {
    const lemmas = [];
    if (w.endsWith("ies")) lemmas.push(w.slice(0, -3) + "y"); // parties->party
    if (w.endsWith("es")) lemmas.push(w.slice(0, -2)); // boxes->box
    lemmas.push(w.slice(0, -1)); // houses->house, cats->cat
    return { ending: "plural", lemmas: dedupe(lemmas) };
  }

  return { ending: "base", lemmas: [w] };
}

function dedupe(arr) {
  return [...new Set(arr.filter(Boolean))];
}

// Re-inflect a base-form candidate to match the surface ending.
// `posBucket` is the POS we're treating the slot as (n/v/a/r) — used so that
// "plural" only applies to nouns and "past/gerund" only to verbs.
/**
 * Handles the gnosis inflect To workflow.
 */
export function inflectTo(base, ending, posBucket) {
  switch (ending) {
    case "plural":
      return posBucket === "n" ? pluralize(base) : null;
    case "past":
      return posBucket === "v" ? pastTense(base) : null;
    case "gerund":
      return posBucket === "v" ? gerund(base) : null;
    case "adverb":
      // candidate may already be an adverb (table r-bucket) or an adjective we adverbize
      return adverbize(base);
    case "base":
    default:
      return base;
  }
}

// ---------------------------------------------------------------------------
// POS resolution.
// ---------------------------------------------------------------------------

// Map a caller posHint to a synonym-table bucket key.
const POS_TO_BUCKET = {
  n: "n", noun: "n",
  v: "v", verb: "v",
  a: "a", adj: "a", adjective: "a", s: "a",
  r: "r", adv: "r", adverb: "r",
};

// Which table bucket should we draw from, given the surface ending + hint?
// Inflection ending strongly implies POS; hint disambiguates the rest.
function resolveBucket({ hint, ending, entry }) {
  if (hint && POS_TO_BUCKET[hint]) return POS_TO_BUCKET[hint];
  if (ending === "plural") return "n";
  if (ending === "past" || ending === "gerund") return "v";
  if (ending === "adverb") return "r"; // adverb slot: r-bucket preferred (adjectives get adverbized)
  // base form, no hint: only safe if the lemma entry has exactly one bucket.
  if (entry) {
    const buckets = Object.keys(entry);
    if (buckets.length === 1) return buckets[0];
  }
  return null; // ambiguous — caller must hint
}

// ---------------------------------------------------------------------------
// Main gate.
// ---------------------------------------------------------------------------

// fluentCandidates(word, posHint?) -> string[]
// Returns grammatically-safe, correctly-inflected synonym surface forms for
// `word`. Empty array => no fluent swap available (caller keeps the original).
/**
 * Handles the gnosis fluent Candidates workflow.
 */
export function fluentCandidates(word, posHint = null) {
  const tbl = table();
  const surface = String(word);
  const { ending, lemmas } = analyzeSurface(surface);

  // Find the first lemma that actually has an entry.
  let lemma = null, entry = null;
  for (const cand of lemmas) {
    if (tbl[cand]) { lemma = cand; entry = tbl[cand]; break; }
  }
  // Fall back to the surface itself (e.g. "difficult" is its own key).
  if (!entry && tbl[surface.toLowerCase()]) {
    lemma = surface.toLowerCase();
    entry = tbl[lemma];
  }
  if (!entry) return [];

  const bucket = resolveBucket({ hint: posHint, ending, entry });
  if (!bucket) return []; // ambiguous base form with no hint -> refuse (safe)

  // For an adverb slot we may pull from BOTH the r-bucket (already adverbs)
  // and the a-bucket (adjectives we adverbize) — but NEVER drop a bare
  // adjective into an adverb slot without the -ly transform.
  let rawCands = [];
  if (ending === "adverb") {
    rawCands = [
      ...(entry.r || []),
      ...(entry.a || []),
    ];
  } else {
    rawCands = entry[bucket] || [];
  }

  const seen = new Set();
  const out = [];
  for (const cand of rawCands) {
    const c = String(cand).toLowerCase();
    // length / identity filter (on the BASE form before inflection)
    if (c.length < 3) continue;
    if (c === lemma || c === surface.toLowerCase()) continue;

    const inflected = inflectTo(c, ending, ending === "adverb" ? "r" : bucket);
    if (!inflected) continue;
    if (inflected.length < 3) continue;
    if (inflected === surface.toLowerCase()) continue; // identity after inflection
    if (seen.has(inflected)) continue;
    seen.add(inflected);
    out.push(inflected);
  }
  return out;
}

// Convenience: raw (ungated) synonyms across all buckets, for comparison/logging.
/**
 * Handles the gnosis raw Candidates workflow.
 */
export function rawCandidates(word) {
  const tbl = table();
  const { lemmas } = analyzeSurface(String(word));
  let entry = null;
  for (const cand of lemmas) if (tbl[cand]) { entry = tbl[cand]; break; }
  if (!entry && tbl[String(word).toLowerCase()]) entry = tbl[String(word).toLowerCase()];
  if (!entry) return [];
  return Object.entries(entry).flatMap(([pos, arr]) => arr.map((w) => `${w}[${pos}]`));
}

// Helper a preconditioner would call: pick the first fluent swap, else keep word.
/**
 * Handles the gnosis precondition workflow.
 */
export function precondition(word, posHint = null) {
  const cands = fluentCandidates(word, posHint);
  return cands.length ? cands[0] : word;
}

// ---------------------------------------------------------------------------
// Self-test (run: `node fluency-gate.mjs`)
// ---------------------------------------------------------------------------

function isMain() {
  return import.meta.url === `file://${process.argv[1]}`;
}

if (isMain()) {
  const cases = [
    ["difficult", "a", "adjective"],
    ["happy", "a", "adjective"],
    ["completed", "v", "verb (past)"],
    ["quickly", "r", "adverb"],
    ["houses", "n", "noun (plural)"],
    ["running", "v", "verb (gerund)"],
  ];

  console.log("=".repeat(72));
  console.log("FLUENCY GATE — raw vs. fluent (filtered + inflected)");
  console.log("=".repeat(72));

  for (const [word, hint, label] of cases) {
    const raw = rawCandidates(word);
    const fluent = fluentCandidates(word, hint);
    const { ending, lemmas } = analyzeSurface(word);
    console.log(`\n● ${word}  (${label})  hint=${hint}  ending=${ending}  lemma=${lemmas[0]}`);
    console.log(`    RAW    [${raw.length}]: ${raw.join(", ") || "—"}`);
    console.log(`    FLUENT [${fluent.length}]: ${fluent.join(", ") || "—"}`);

    // Annotate what got dropped and why.
    const dropped = [];
    for (const r of raw) {
      const m = r.match(/^(.*)\[(.)\]$/);
      const base = m[1], pos = m[2];
      const inSlot =
        ending === "adverb" ? (pos === "r" || pos === "a")
        : ending === "plural" ? pos === "n"
        : (ending === "past" || ending === "gerund") ? pos === "v"
        : true;
      if (!inSlot) dropped.push(`${base}[${pos}] (POS mismatch)`);
    }
    if (dropped.length) console.log(`    DROPPED (POS): ${dropped.join(", ")}`);
  }
  console.log("\n" + "=".repeat(72));
}
