// @a0n/.ignored_gnosis/forkjoin-app/src/terminology.ts
/**
 * Shared product terminology across forkjoin-app surfaces.
 * Keep these labels stable so copy stays consistent across routes.
 */

export const TERM_STARGATE = 'Stargate';
export const TERM_COMMUNICATION_LAYER = 'Communication Layer';
export const TERM_STORE = 'Store';
export const TERM_ZEITGEIST = 'Zeitgeist';
export const TERM_COVENANT = 'Covenant';
export const TERM_CRYPTO_RELAY = 'Crypto Relay';
export const TERM_DASHRELAY = 'dashrelay';
export const TERM_DASH = 'Dash';
export const TERM_RELAY = 'Relay';

export const STARGATE_CHAT_AND_COVENANT_LABEL = `${TERM_STARGATE} (Chat + ${TERM_COVENANT})`;
export const STARGATE_STACK_DESCRIPTION = `${TERM_STARGATE} is the front door to Nexus conversations, ${TERM_ZEITGEIST} discovery, ${TERM_COVENANT} flows, and ${TERM_STORE} apps.`;

export const DASHRELAY_TRUST_LOG_DESCRIPTION =
  'Messaging + CRDT broker for local-first apps. UCAN identity, zero-knowledge encryption, and an invisible verifiable trust log.';

export const STARGATE_WEB_HOST = 'app.forkjoin.ai';
export const STARGATE_WEB_TITLE = 'ForkJoin Stargate';
export const STARGATE_WEB_SUBTITLE =
  'Discover phyles and bring them online at the edge.';
