// @a0n/.ignored_gnosis/forkjoin-app/src/worker.ts
import { Hono } from '@a0n/x-gnosis/compat/hono';
import { cors } from '@a0n/x-gnosis/compat/hono';
import {
  buildPublicSkillsIndex,
  buildSkillResources,
  buildSkillsListPayload,
  getSkillByName,
  readSkillResource,
  renderSkillMarkdown,
} from '@a0n/shared-utils/services/skills';
import { GnosisDebugSessionDurableObject } from './durables/GnosisDebugSessionDO';
import { GnosisMcpSessionDurableObject } from './durables/GnosisMcpSessionDO';
import {
  AUTH_TOOL_NAMES,
  GNOSIS_MCP_TOOLS,
  PUBLIC_TOOL_NAMES,
  consumePublicToolQuota,
  getToolByName,
  invokeGnosisTool,
} from './mcp/tools';
import { GNOSIS_MCP_RESOURCES, readGnosisResource } from './mcp/resources';
import {
  GNOSIS_SKILLS,
  buildAgentCardManifest,
  buildAgentsManifest,
  buildAiPluginManifest,
  buildDiscoveryEndpoints,
  buildLlmsFullTxt,
  buildLlmsTxt,
  buildMcpMarketingManifest,
  buildMcpWellKnownManifest,
  buildOpenApiManifest,
  buildSecurityTxt,
  resolvePublicBaseUrl,
} from './mcp/manifests';
import type { Env, McpRequest, McpResponse, ToolCallResult } from './types';

export { GnosisMcpSessionDurableObject, GnosisDebugSessionDurableObject };

const MCP_PROTOCOL_VERSION = '2024-11-05';
const MCP_SERVER_NAME = 'forkjoin-app';
const MCP_SERVER_VERSION = '0.1.0';
const MCP_SESSION_HEADER = 'mcp-session-id';

const DISCOVERY_CACHE_CONTROL = 'public, max-age=3600, s-maxage=86400';

const WEB_ROUTES = new Set([
  '/',
  '/ide',
  '/repl',
  '/playground',
  '/docs',
  '/privacy',
  '/cookies',
  '/cookie-settings',
  '/colophon',
  '/terms',
]);

const app = new Hono<{ Bindings: Env }>();

app.use(
  '*',
  cors({
    origin: '*',
    allowMethods: ['GET', 'POST', 'OPTIONS'],
    allowHeaders: ['Content-Type', 'Authorization', MCP_SESSION_HEADER],
    exposeHeaders: [MCP_SESSION_HEADER],
    maxAge: 86400,
  })
);

app.get('/health', async (c) => {
  const sessionId = parseSessionId(c.req.raw);
  let sessionSummary: unknown = null;

  if (sessionId) {
    sessionSummary = await getSessionSummarySafe(c.env, sessionId);
  }

  return json(
    {
      status: 'healthy',
      service: MCP_SERVER_NAME,
      version: MCP_SERVER_VERSION,
      environment: c.env.ENVIRONMENT,
      timestamp: new Date().toISOString(),
      mcp: {
        endpoint: '/mcp',
        sessionHeader: MCP_SESSION_HEADER,
        toolCount: GNOSIS_MCP_TOOLS.length,
        publicTools: Array.from(PUBLIC_TOOL_NAMES).sort(),
        authTools: Array.from(AUTH_TOOL_NAMES).sort(),
      },
      relay: {
        baseUrl: c.env.DASH_RELAY_URL ?? 'https://relay.dashrelay.com',
        ws: c.env.DASH_RELAY_WS_URL ?? 'wss://relay.dashrelay.com/relay/sync',
        webtransport:
          c.env.DASH_RELAY_WT_URL ?? 'https://relay.dashrelay.com/relay',
        discovery:
          c.env.DASH_RELAY_DISCOVERY_URL ??
          'https://relay.dashrelay.com/discovery',
      },
      session: sessionSummary,
    },
    {
      cacheControl: 'public, max-age=30, s-maxage=60',
    }
  );
});

app.get('/session', async (c) => {
  const sessionId = parseSessionId(c.req.raw);
  if (!sessionId) {
    return json(
      {
        ok: false,
        error: 'missing_session_id',
        message: `Set ${MCP_SESSION_HEADER} header to inspect session state.`,
      },
      {
        status: 400,
      }
    );
  }

  const sessionIdObject = c.env.MCP_SESSION_DO.idFromName(sessionId);
  const stub = c.env.MCP_SESSION_DO.get(sessionIdObject);
  const response = await stub.fetch(
    new Request('https://gnosis-mcp.internal/summary')
  );
  const body = await response.text();

  return new Response(body, {
    status: response.status,
    headers: {
      'content-type':
        response.headers.get('content-type') ?? 'application/json',
      'cache-control': 'no-store',
      [MCP_SESSION_HEADER]: sessionId,
    },
  });
});

app.get('/mcp', async (c) => {
  const endpoints = buildDiscoveryEndpoints(
    c.req.raw,
    c.env.MCP_PUBLIC_BASE_URL
  );
  return json(
    {
      name: MCP_SERVER_NAME,
      version: MCP_SERVER_VERSION,
      protocolVersion: MCP_PROTOCOL_VERSION,
      endpoint: endpoints.mcp,
      initializeMethod: 'initialize',
      toolsListMethod: 'tools/list',
      toolsCallMethod: 'tools/call',
      resourcesListMethod: 'resources/list',
      resourcesReadMethod: 'resources/read',
      sessionHeader: MCP_SESSION_HEADER,
      discovery: {
        llms: endpoints.llms,
        llmsFull: endpoints.llmsFull,
        skills: endpoints.skills,
        skillsIndex: endpoints.skillsIndex,
        agents: endpoints.agents,
        mcp: endpoints.mcpWellKnown,
        mcpManifest: endpoints.mcpManifest,
        agentCard: endpoints.agentCard,
        openApi: endpoints.openApi,
        security: endpoints.security,
      },
    },
    {
      cacheControl: DISCOVERY_CACHE_CONTROL,
    }
  );
});

app.post('/mcp', async (c) => {
  const body = await parseBody(c.req.raw);
  if (!body.ok) {
    return json(mcpError(null, -32700, body.error), {
      status: 400,
    });
  }

  const parsedRequest = parseMcpRequest(body.value);
  if (!parsedRequest.ok) {
    return json(parsedRequest.error, {
      status: 400,
    });
  }

  const sessionId = parseSessionId(c.req.raw);
  const authToken = parseAuthToken(c.req.raw);

  try {
    const result = await handleMcpRequest(
      parsedRequest.request,
      c.env,
      c.req.raw,
      sessionId,
      authToken
    );

    return json(result.payload, {
      sessionId: result.sessionId,
    });
  } catch (error) {
    return json(
      mcpError(parsedRequest.request.id, -32603, 'Internal error', {
        message: error instanceof Error ? error.message : 'Unknown error',
      }),
      {
        status: 500,
        sessionId,
      }
    );
  }
});

app.get('/llms.txt', (c) => {
  const endpoints = buildDiscoveryEndpoints(
    c.req.raw,
    c.env.MCP_PUBLIC_BASE_URL
  );
  return text(buildLlmsTxt(endpoints), {
    cacheControl: DISCOVERY_CACHE_CONTROL,
  });
});

app.get('/llms-full.txt', (c) => {
  const endpoints = buildDiscoveryEndpoints(
    c.req.raw,
    c.env.MCP_PUBLIC_BASE_URL
  );
  return text(buildLlmsFullTxt(endpoints), {
    cacheControl: DISCOVERY_CACHE_CONTROL,
  });
});

app.get('/skills', (c) => {
  const endpoints = buildDiscoveryEndpoints(
    c.req.raw,
    c.env.MCP_PUBLIC_BASE_URL
  );
  return json(
    buildSkillsListPayload({
      server: {
        id: MCP_SERVER_NAME,
        name: 'ForkJoin MCP',
        version: MCP_SERVER_VERSION,
      },
      skills: GNOSIS_SKILLS,
      discovery: {
        endpoint: endpoints.mcp,
        llms: endpoints.llms,
        manifest: endpoints.mcpWellKnown,
        agentCard: endpoints.agentCard,
      },
    }),
    {
      cacheControl: DISCOVERY_CACHE_CONTROL,
    }
  );
});

app.get('/skills.json', (c) => {
  const endpoints = buildDiscoveryEndpoints(
    c.req.raw,
    c.env.MCP_PUBLIC_BASE_URL
  );
  return json(
    buildSkillsListPayload({
      server: {
        id: MCP_SERVER_NAME,
        name: 'ForkJoin MCP',
        version: MCP_SERVER_VERSION,
      },
      skills: GNOSIS_SKILLS,
      discovery: {
        endpoint: endpoints.mcp,
        llms: endpoints.llms,
        manifest: endpoints.mcpWellKnown,
        agentCard: endpoints.agentCard,
      },
    }),
    {
      cacheControl: DISCOVERY_CACHE_CONTROL,
    }
  );
});

app.get('/skills/index.json', (c) => {
  const endpoints = buildDiscoveryEndpoints(
    c.req.raw,
    c.env.MCP_PUBLIC_BASE_URL
  );
  return json(
    buildPublicSkillsIndex({
      server: {
        id: MCP_SERVER_NAME,
        name: 'ForkJoin MCP',
        version: MCP_SERVER_VERSION,
      },
      discovery: {
        endpoint: endpoints.mcp,
        llms: endpoints.llms,
        manifest: endpoints.mcpWellKnown,
        agentCard: endpoints.agentCard,
      },
      skills: GNOSIS_SKILLS,
      featuredSkills: GNOSIS_SKILLS.slice(0, 2).map((skill) => skill.id),
    }),
    {
      cacheControl: DISCOVERY_CACHE_CONTROL,
    }
  );
});

app.get('/skills/:name.md', (c) => {
  const skillParam = c.req.param('name');
  if (!skillParam) {
    return json(
      {
        ok: false,
        error: 'bad_request',
        message: 'Missing skill name parameter.',
      },
      {
        status: 400,
      }
    );
  }

  const skillName = decodeURIComponent(skillParam);
  const skill = getSkillByName(GNOSIS_SKILLS, skillName);

  if (!skill) {
    return json(
      {
        ok: false,
        error: 'not_found',
        message: `Unknown skill: ${skillName}`,
      },
      {
        status: 404,
      }
    );
  }

  return text(renderSkillMarkdown(skill), {
    cacheControl: DISCOVERY_CACHE_CONTROL,
    contentType: 'text/markdown; charset=utf-8',
  });
});

app.get('/agents.json', (c) => {
  const endpoints = buildDiscoveryEndpoints(
    c.req.raw,
    c.env.MCP_PUBLIC_BASE_URL
  );
  return json(buildAgentsManifest(endpoints), {
    cacheControl: DISCOVERY_CACHE_CONTROL,
  });
});

app.get('/.well-known/agents.json', (c) => {
  const endpoints = buildDiscoveryEndpoints(
    c.req.raw,
    c.env.MCP_PUBLIC_BASE_URL
  );
  return json(buildAgentsManifest(endpoints), {
    cacheControl: DISCOVERY_CACHE_CONTROL,
  });
});

app.get('/.well-known/ai-plugin.json', (c) => {
  const endpoints = buildDiscoveryEndpoints(
    c.req.raw,
    c.env.MCP_PUBLIC_BASE_URL
  );
  return json(buildAiPluginManifest(endpoints), {
    cacheControl: DISCOVERY_CACHE_CONTROL,
  });
});

app.get('/.well-known/mcp.json', (c) => {
  const endpoints = buildDiscoveryEndpoints(
    c.req.raw,
    c.env.MCP_PUBLIC_BASE_URL
  );
  return json(buildMcpWellKnownManifest(endpoints), {
    cacheControl: DISCOVERY_CACHE_CONTROL,
  });
});

app.get('/.well-known/mcp-manifest.json', (c) => {
  const endpoints = buildDiscoveryEndpoints(
    c.req.raw,
    c.env.MCP_PUBLIC_BASE_URL
  );
  return json(buildMcpMarketingManifest(endpoints), {
    cacheControl: DISCOVERY_CACHE_CONTROL,
  });
});

app.get('/.well-known/agent-card.json', (c) => {
  const endpoints = buildDiscoveryEndpoints(
    c.req.raw,
    c.env.MCP_PUBLIC_BASE_URL
  );
  return json(buildAgentCardManifest(endpoints), {
    cacheControl: DISCOVERY_CACHE_CONTROL,
  });
});

app.get('/.well-known/openapi.json', (c) => {
  const endpoints = buildDiscoveryEndpoints(
    c.req.raw,
    c.env.MCP_PUBLIC_BASE_URL
  );
  return json(buildOpenApiManifest(endpoints), {
    cacheControl: DISCOVERY_CACHE_CONTROL,
  });
});

app.get('/.well-known/security.txt', (c) => {
  const endpoints = buildDiscoveryEndpoints(
    c.req.raw,
    c.env.MCP_PUBLIC_BASE_URL
  );
  return text(buildSecurityTxt(endpoints), {
    cacheControl: 'public, max-age=86400, s-maxage=86400',
  });
});

app.get('/*', async (c) => {
  const path = new URL(c.req.url).pathname;
  const assetResponse = await fetchAssetRoute(c.req.raw, c.env);

  if (assetResponse) {
    return assetResponse;
  }

  if (WEB_ROUTES.has(path)) {
    return new Response(buildDevelopmentFallbackHtml(path), {
      status: 200,
      headers: {
        'content-type': 'text/html; charset=utf-8',
        'cache-control': 'no-store',
      },
    });
  }

  return json(
    {
      ok: false,
      error: 'not_found',
      message: `Unknown route: ${path}`,
      knownWebRoutes: Array.from(WEB_ROUTES),
    },
    {
      status: 404,
    }
  );
});

interface ParsedMcpRequest {
  ok: true;
  request: McpRequest;
}

interface FailedMcpRequest {
  ok: false;
  error: McpResponse;
}

interface McpExecutionResult {
  payload: McpResponse;
  sessionId: string | null;
}

interface ToolAuthorizationDecision {
  allowed: boolean;
  reason: string | null;
  requiresAuth: boolean;
}

/**
 * Parses the Session Id.
 */
export function parseSessionId(request: Request): string | null {
  const raw = request.headers.get(MCP_SESSION_HEADER);
  if (!raw) {
    return null;
  }

  const normalized = raw.trim();
  return normalized.length > 0 ? normalized : null;
}

/**
 * Parses the Auth Token.
 */
export function parseAuthToken(request: Request): string | null {
  const raw = request.headers.get('authorization');
  if (!raw) {
    return null;
  }

  const [scheme, token] = raw.trim().split(/\s+/, 2);
  if (!scheme || !token || scheme.toLowerCase() !== 'bearer') {
    return null;
  }

  const normalized = token.trim();
  return normalized.length > 0 ? normalized : null;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

async function parseBody(
  request: Request
): Promise<{ ok: true; value: unknown } | { ok: false; error: string }> {
  try {
    const value = (await request.json()) as unknown;
    return { ok: true, value };
  } catch {
    return { ok: false, error: 'Parse error: request body must be valid JSON' };
  }
}

/**
 * Parses the Mcp Request.
 */
export function parseMcpRequest(
  body: unknown
): ParsedMcpRequest | FailedMcpRequest {
  if (!isRecord(body)) {
    return {
      ok: false,
      error: mcpError(null, -32600, 'Invalid Request: expected JSON object'),
    };
  }

  const jsonrpc = body.jsonrpc;
  const id = body.id;
  const method = body.method;

  if (
    jsonrpc !== '2.0' ||
    typeof method !== 'string' ||
    method.trim().length === 0
  ) {
    return {
      ok: false,
      error: mcpError(
        typeof id === 'string' || typeof id === 'number' || id === null
          ? id
          : null,
        -32600,
        'Invalid Request: jsonrpc must be "2.0" and method must be a non-empty string'
      ),
    };
  }

  return {
    ok: true,
    request: {
      jsonrpc: '2.0',
      id:
        typeof id === 'string' || typeof id === 'number' || id === null
          ? id
          : null,
      method: method.trim(),
      params: isRecord(body.params) ? body.params : undefined,
    },
  };
}

function mcpError(
  id: string | number | null,
  code: number,
  message: string,
  data?: unknown
): McpResponse {
  return {
    jsonrpc: '2.0',
    id,
    error: {
      code,
      message,
      ...(data === undefined ? {} : { data }),
    },
  };
}

interface ToolResultPayload {
  content: Array<{
    type: 'text';
    text: string;
  }>;
  isError?: true;
}

function stringifyToolResult(result: ToolCallResult): ToolResultPayload {
  return {
    content: result.content,
    ...(result.isError ? { isError: true as const } : {}),
  };
}

/**
 * Resolves the Auth Token Set.
 */
export function resolveAuthTokenSet(env: Env): Set<string> {
  const configured = new Set<string>();

  if (typeof env.MCP_AUTH_TOKEN === 'string') {
    const token = env.MCP_AUTH_TOKEN.trim();
    if (token.length > 0) {
      configured.add(token);
    }
  }

  if (typeof env.MCP_AUTH_TOKENS === 'string') {
    env.MCP_AUTH_TOKENS.split(',')
      .map((token) => token.trim())
      .filter((token) => token.length > 0)
      .forEach((token) => configured.add(token));
  }

  return configured;
}

function isAuthTokenValid(token: string | null, env: Env): boolean {
  const configured = resolveAuthTokenSet(env);
  if (configured.size === 0) {
    return false;
  }

  if (!token) {
    return false;
  }

  return configured.has(token);
}

/**
 * Resolves the Tool Authorization.
 */
export function resolveToolAuthorization(
  toolName: string,
  token: string | null,
  env: Env
): ToolAuthorizationDecision {
  if (PUBLIC_TOOL_NAMES.has(toolName)) {
    return {
      allowed: true,
      reason: null,
      requiresAuth: false,
    };
  }

  if (!AUTH_TOOL_NAMES.has(toolName)) {
    return {
      allowed: false,
      reason: `Unknown tool: ${toolName}`,
      requiresAuth: false,
    };
  }

  if (isAuthTokenValid(token, env)) {
    return {
      allowed: true,
      reason: null,
      requiresAuth: true,
    };
  }

  return {
    allowed: false,
    reason:
      'This tool requires bearer auth. Set MCP_AUTH_TOKEN in request headers.',
    requiresAuth: true,
  };
}

async function initializeSession(env: Env, sessionId: string): Promise<void> {
  const sessionIdObject = env.MCP_SESSION_DO.idFromName(sessionId);
  const stub = env.MCP_SESSION_DO.get(sessionIdObject);
  await stub.fetch(
    new Request('https://gnosis-mcp.internal/initialize', {
      method: 'POST',
      headers: {
        'content-type': 'application/json',
      },
      body: JSON.stringify({ sessionId }),
    })
  );
}

async function recordSessionEvent(
  env: Env,
  sessionId: string | null,
  eventType: string,
  detail?: Record<string, unknown>
): Promise<void> {
  if (!sessionId) {
    return;
  }

  const sessionIdObject = env.MCP_SESSION_DO.idFromName(sessionId);
  const stub = env.MCP_SESSION_DO.get(sessionIdObject);
  await stub.fetch(
    new Request('https://gnosis-mcp.internal/record', {
      method: 'POST',
      headers: {
        'content-type': 'application/json',
      },
      body: JSON.stringify({
        type: eventType,
        detail: detail ?? {},
      }),
    })
  );
}

async function getSessionSummarySafe(
  env: Env,
  sessionId: string
): Promise<unknown> {
  try {
    const sessionIdObject = env.MCP_SESSION_DO.idFromName(sessionId);
    const stub = env.MCP_SESSION_DO.get(sessionIdObject);
    const response = await stub.fetch(
      new Request('https://gnosis-mcp.internal/summary')
    );
    return (await response.json()) as unknown;
  } catch {
    return {
      ok: false,
      error: 'session_unavailable',
    };
  }
}

async function handleMcpRequest(
  request: McpRequest,
  env: Env,
  rawRequest: Request,
  existingSessionId: string | null,
  authToken: string | null
): Promise<McpExecutionResult> {
  const { id, method } = request;
  const params = isRecord(request.params) ? request.params : {};

  switch (method) {
    case 'initialize': {
      const sessionId = existingSessionId ?? crypto.randomUUID();
      await initializeSession(env, sessionId);
      await recordSessionEvent(env, sessionId, 'initialize', {
        protocolVersion: MCP_PROTOCOL_VERSION,
      });

      return {
        sessionId,
        payload: {
          jsonrpc: '2.0',
          id,
          result: {
            protocolVersion: MCP_PROTOCOL_VERSION,
            serverInfo: {
              name: MCP_SERVER_NAME,
              version: MCP_SERVER_VERSION,
            },
            capabilities: {
              tools: {
                listChanged: true,
              },
              resources: {
                listChanged: true,
              },
            },
          },
        },
      };
    }

    case 'tools/list': {
      await recordSessionEvent(env, existingSessionId, 'tools/list');
      return {
        sessionId: existingSessionId,
        payload: {
          jsonrpc: '2.0',
          id,
          result: {
            tools: GNOSIS_MCP_TOOLS,
          },
        },
      };
    }

    case 'tools/call': {
      const toolNameRaw = params.name;
      if (typeof toolNameRaw !== 'string' || toolNameRaw.trim().length === 0) {
        return {
          sessionId: existingSessionId,
          payload: mcpError(id, -32602, 'Tool name is required.'),
        };
      }

      const toolName = toolNameRaw.trim();
      if (!getToolByName(toolName)) {
        return {
          sessionId: existingSessionId,
          payload: mcpError(id, -32602, `Unknown tool: ${toolName}`),
        };
      }

      const authDecision = resolveToolAuthorization(toolName, authToken, env);
      if (!authDecision.allowed) {
        return {
          sessionId: existingSessionId,
          payload: mcpError(
            id,
            -32001,
            authDecision.reason ?? 'Not authorized',
            {
              toolName,
              requiresAuth: authDecision.requiresAuth,
            }
          ),
        };
      }

      const toolArgs = isRecord(params.arguments)
        ? { ...params.arguments }
        : {};

      let quota: {
        allowed: boolean;
        remaining: number;
        limitPerHour: number;
      } | null = null;
      const sessionId = existingSessionId ?? crypto.randomUUID();
      await initializeSession(env, sessionId);

      if (PUBLIC_TOOL_NAMES.has(toolName)) {
        quota = await consumePublicToolQuota(env, sessionId, toolName);

        if (!quota.allowed) {
          return {
            sessionId,
            payload: mcpError(
              id,
              -32011,
              'Free tier quota exceeded. Subscribe for $3/month for unlimited access.',
              {
                toolName,
                limitPerDay: quota.limitPerHour,
                remaining: quota.remaining,
                subscribeUrl: 'https://forkjoin.ai/#/subscribe',
              }
            ),
          };
        }
      }

      const toolResult = await invokeGnosisTool(toolName, toolArgs, {
        env,
        sessionId,
        authToken,
      });

      await recordSessionEvent(env, sessionId, 'tools/call', {
        toolName,
        access: PUBLIC_TOOL_NAMES.has(toolName) ? 'public' : 'auth',
        ok: !toolResult.isError,
      });

      return {
        sessionId,
        payload: {
          jsonrpc: '2.0',
          id,
          result: {
            ...stringifyToolResult(toolResult),
            ...(quota
              ? {
                  quota: {
                    limitPerDay: quota.limitPerHour,
                    remaining: quota.remaining,
                  },
                }
              : {}),
          },
        },
      };
    }

    case 'resources/list': {
      await recordSessionEvent(env, existingSessionId, 'resources/list');
      return {
        sessionId: existingSessionId,
        payload: {
          jsonrpc: '2.0',
          id,
          result: {
            resources: [
              ...GNOSIS_MCP_RESOURCES,
              ...buildSkillResources(GNOSIS_SKILLS),
            ],
          },
        },
      };
    }

    case 'resources/read': {
      const uriRaw = params.uri;
      if (typeof uriRaw !== 'string' || uriRaw.trim().length === 0) {
        return {
          sessionId: existingSessionId,
          payload: mcpError(id, -32602, 'Resource URI is required.'),
        };
      }

      const uri = uriRaw.trim();
      const skillResource = readSkillResource(GNOSIS_SKILLS, uri);
      if (skillResource) {
        await recordSessionEvent(env, existingSessionId, 'resources/read', {
          uri,
        });
        return {
          sessionId: existingSessionId,
          payload: {
            jsonrpc: '2.0',
            id,
            result: {
              contents: [
                {
                  uri: skillResource.uri,
                  mimeType: skillResource.mimeType,
                  text: skillResource.text,
                },
              ],
            },
          },
        };
      }

      try {
        const publicBaseUrl = resolvePublicBaseUrl(
          rawRequest,
          env.MCP_PUBLIC_BASE_URL
        );
        const resource = readGnosisResource(uri, env, publicBaseUrl);
        await recordSessionEvent(env, existingSessionId, 'resources/read', {
          uri,
        });
        return {
          sessionId: existingSessionId,
          payload: {
            jsonrpc: '2.0',
            id,
            result: resource,
          },
        };
      } catch (error) {
        return {
          sessionId: existingSessionId,
          payload: mcpError(
            id,
            -32602,
            error instanceof Error ? error.message : 'Unknown resource error'
          ),
        };
      }
    }

    case 'ping': {
      await recordSessionEvent(env, existingSessionId, 'ping');
      return {
        sessionId: existingSessionId,
        payload: {
          jsonrpc: '2.0',
          id,
          result: {
            ok: true,
            now: new Date().toISOString(),
          },
        },
      };
    }

    default: {
      return {
        sessionId: existingSessionId,
        payload: mcpError(id, -32601, `Method not found: ${method}`),
      };
    }
  }
}

async function fetchAssetRoute(
  request: Request,
  env: Env
): Promise<Response | null> {
  if (!env.ASSETS) {
    return null;
  }

  try {
    const direct = await env.ASSETS.fetch(request);
    if (direct.status !== 404) {
      return direct;
    }

    const indexRequest = new Request(
      new URL('/index.html', request.url).toString(),
      {
        method: 'GET',
        headers: request.headers,
      }
    );
    const indexResponse = await env.ASSETS.fetch(indexRequest);
    if (indexResponse.status === 404) {
      return null;
    }

    return indexResponse;
  } catch {
    return null;
  }
}

function buildDevelopmentFallbackHtml(path: string): string {
  return `<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <title>ForkJoin (Development)</title>
    <style>
      body {
        margin: 0;
        min-height: 100vh;
        display: grid;
        place-items: center;
        font-family: ui-sans-serif, system-ui, -apple-system, Segoe UI, Roboto, sans-serif;
        background: radial-gradient(circle at top, #163047 0%, #050910 58%, #020306 100%);
        color: #dce7f5;
      }
      .card {
        max-width: 720px;
        border: 1px solid rgba(173, 208, 255, 0.22);
        background: rgba(8, 16, 28, 0.78);
        padding: 24px;
        border-radius: 16px;
      }
      code { color: #9bd4ff; }
      a { color: #8de7ff; }
    </style>
  </head>
  <body>
    <article class="card">
      <h1>ForkJoin Worker Running</h1>
      <p>Route <code>${escapeHtml(
        path
      )}</code> is available, but static assets are not bound in this environment.</p>
      <p>Run <code>bun run build:client</code> and bind <code>ASSETS</code>, or use Vite dev server at <code>http://127.0.0.1:5196</code>.</p>
      <p>MCP endpoint is live at <a href="/mcp">/mcp</a>.</p>
    </article>
  </body>
</html>`;
}

function escapeHtml(input: string): string {
  return input
    .replaceAll('&', '&amp;')
    .replaceAll('<', '&lt;')
    .replaceAll('>', '&gt;')
    .replaceAll('"', '&quot;')
    .replaceAll("'", '&#39;');
}

function json(
  payload: unknown,
  options: {
    status?: number;
    sessionId?: string | null;
    cacheControl?: string;
  } = {}
): Response {
  const headers = new Headers({
    'content-type': 'application/json; charset=utf-8',
    'cache-control': options.cacheControl ?? 'no-store',
  });

  if (options.sessionId) {
    headers.set(MCP_SESSION_HEADER, options.sessionId);
  }

  return new Response(JSON.stringify(payload), {
    status: options.status ?? 200,
    headers,
  });
}

function text(
  payload: string,
  options: { status?: number; cacheControl?: string; contentType?: string } = {}
): Response {
  return new Response(payload, {
    status: options.status ?? 200,
    headers: {
      'content-type': options.contentType ?? 'text/plain; charset=utf-8',
      'cache-control': options.cacheControl ?? 'no-store',
    },
  });
}

export default app;
