// @a0n/.ignored_gnosis/src/betti/polyglot-betti-bridge.ts
import * as path from 'node:path';
import * as child_process from 'node:child_process';
import { fileURLToPath } from 'node:url';

/**
 * polyglot-betti-bridge.ts
 *
 * JS-side half of the polyglot WASM pipeline.
 *
 * Architecture:
 *   JS  →  web-tree-sitter or hand-built extraction  →  PolyglotFnInput JSON
 *   WASM →  compile_polyglot_fn(json)                 →  PolyglotGnosisResult
 *   Native → gnosis-betti binary (optional)          →  PolyglotGnosisResult
 *
 * tree-sitter C grammars cannot link to wasm32-unknown-unknown (no libc).
 * Parsing therefore stays here in JS; topology compilation stays in Rust WASM.
 *
 * Usage:
 *   import { buildPolyglotFnInput, compilePolyglotFn } from './polyglot-betti-bridge';
 *   const input = buildPolyglotFnInput('python', 'process_data', ['data'], body);
 *   const result = await compilePolyglotFn(input);
 */

// ── Wire-format types (mirror polyglot_wasm_types.rs exactly) ─────────────

export type PolyglotArgInput =
  | { kind: 'identifier'; name: string; ghost_mass?: any }
  | { kind: 'literal'; value: unknown; ghost_mass?: any }
  | { kind: 'member'; target: PolyglotArgInput; property: string; ghost_mass?: any }
  | { kind: 'call'; callee: PolyglotArgInput; args: PolyglotArgInput[]; ghost_mass?: any };

export type PolyglotCallInput = {
  callee: PolyglotArgInput;
  args: PolyglotArgInput[];
  assign_to?: string;
  ghost_mass?: any;
};

export type PolyglotStmtInput =
  | { kind: 'var_decl'; bindings: string[]; init?: PolyglotArgInput; ghost_mass?: any }
  | { kind: 'call'; callee: PolyglotArgInput; args: PolyglotArgInput[]; ghost_mass?: any }
  | { kind: 'return'; value?: PolyglotArgInput; ghost_mass?: any }
  | { kind: 'parallel'; bindings: string[]; branches: PolyglotCallInput[]; ghost_mass?: any };

export type PolyglotFnInput = {
  name: string;
  params: string[];
  body: PolyglotStmtInput[];
  language?: string;
  file_path?: string;
};

// ── Result types ───────────────────────────────────────────────────────────

export type PolyglotGnosisResult = {
  functionName: string;
  language: string;
  sourceFilePath: string;
  entryNodeId: string;
  ggSource: string;
  topologySource: string;
  ast: { nodes: unknown[]; edges: unknown[] };
  nodePlans: unknown[];
  schedule: unknown[];
  transpiledSources: Record<string, string>;
};

// ── WASM loading ───────────────────────────────────────────────────────────

let wasmModule: {
  compile_polyglot_fn(json: string): PolyglotGnosisResult;
  polyglot_supported_languages(): string[];
} | null = null;

/**
 * Load the gnosis-betti WASM module.
 * Must be called before `compilePolyglotFn`.
 */
export async function initPolyglotBettiWasm(
  wasmUrl?: string,
): Promise<void> {
  // Prefer the caller's URL; fall back to the pkg-adjacent default.
  const url = wasmUrl ?? new URL('../../gnosis-betti-wasm/pkg/gnosis_betti_wasm.js', import.meta.url).href;
  const mod = await import(/* @vite-ignore */ url);
  if (typeof mod.default === 'function') {
      await mod.default();
  } else if (mod.default && typeof mod.default.default === 'function') {
      await mod.default.default();
  } else if (typeof mod.initSync === 'function') {
      // noop async fallback
  }
  wasmModule = mod;
}

function requireWasm() {
  if (!wasmModule) {
    throw new Error(
      'gnosis-betti WASM not initialised — call initPolyglotBettiWasm() first',
    );
  }
  return wasmModule;
}

// ── Native execution ───────────────────────────────────────────────────────

function invokeNativeBetti(args: string[], input?: string): any {
  if (typeof process === 'undefined' || !process.env) return null;

  const currentDir = typeof __dirname === 'string' ? __dirname : path.dirname(fileURLToPath(import.meta.url));
  const binPath =
    process.env.GNOSIS_BETTI_BIN ||
    path.resolve(currentDir, '..', '..', 'gnosis-betti-wasm', 'target', 'release', 'gnosis-betti');

  try {
    const result = child_process.spawnSync(binPath, args, {
      input,
      encoding: 'utf8',
      maxBuffer: 1024 * 1024 * 50, // 50MB
    });

    if (result.error || result.status !== 0) {
      return null;
    }

    return JSON.parse(result.stdout);
  } catch {
    return null;
  }
}

// ── Public API ─────────────────────────────────────────────────────────────

/**
 * Compile a pre-extracted polyglot function to a Betti `.gg` topology.
 * The caller is responsible for extracting `body` using web-tree-sitter or
 * any other AST source.
 */
export function compilePolyglotFn(input: PolyglotFnInput): PolyglotGnosisResult {
  const nativeResult = invokeNativeBetti(['compile-polyglot'], JSON.stringify(input));
  if (nativeResult) {
    return nativeResult;
  }

  const wasm = requireWasm();
  return wasm.compile_polyglot_fn(JSON.stringify(input));
}

/**
 * Return the list of languages the polyglot WASM compiler supports.
 */
export function polyglotSupportedLanguages(): string[] {
  const nativeResult = invokeNativeBetti(['languages']);
  if (nativeResult) {
    return nativeResult;
  }

  return requireWasm().polyglot_supported_languages();
}

// ── Builder helpers ────────────────────────────────────────────────────────

/** Convenience: construct a PolyglotFnInput from discrete parts. */
export function buildPolyglotFnInput(
  language: string,
  name: string,
  params: string[],
  body: PolyglotStmtInput[],
  filePath?: string,
): PolyglotFnInput {
  return { name, params, body, language, file_path: filePath };
}

/** Shorthand: identifier arg. */
export const id = (name: string, ghostMass?: any): PolyglotArgInput => ({ kind: 'identifier', name, ghost_mass: ghostMass });

/** Shorthand: literal arg. */
export const lit = (value: unknown, ghostMass?: any): PolyglotArgInput => ({ kind: 'literal', value, ghost_mass: ghostMass });

/** Shorthand: member access arg. */
export const member = (target: PolyglotArgInput, property: string, ghostMass?: any): PolyglotArgInput => ({
  kind: 'member', target, property, ghost_mass: ghostMass
});

/** Shorthand: call arg (for use inside expressions — hoisted by compiler). */
export const callArg = (
  callee: PolyglotArgInput,
  args: PolyglotArgInput[],
  ghostMass?: any
): PolyglotArgInput => ({ kind: 'call', callee, args, ghost_mass: ghostMass });

/** Shorthand: var_decl statement. */
export const varDecl = (
  bindings: string[],
  init?: PolyglotArgInput,
  ghostMass?: any
): PolyglotStmtInput => ({ kind: 'var_decl', bindings, init, ghost_mass: ghostMass });

/** Shorthand: bare call statement (side-effect). */
export const callStmt = (
  callee: PolyglotArgInput,
  args: PolyglotArgInput[],
  ghostMass?: any
): PolyglotStmtInput => ({ kind: 'call', callee, args, ghost_mass: ghostMass });

/** Shorthand: return statement. */
export const ret = (value?: PolyglotArgInput, ghostMass?: any): PolyglotStmtInput => ({
  kind: 'return', value, ghost_mass: ghostMass
});

/** Shorthand: parallel fan-out (Promise.all / asyncio.gather / tokio::join!). */
export const parallel = (
  bindings: string[],
  branches: PolyglotCallInput[],
  ghostMass?: any
): PolyglotStmtInput => ({ kind: 'parallel', bindings, branches, ghost_mass: ghostMass });
