// @a0n/.ignored_gnosis/src/betti/s-duality-router.ts
/**
 * s-duality-router.ts — Langlands S-Duality execution router.
 *
 * The core insight: L_n = F_{n+1} + F_{n-1}.
 * The Lucas sequence IS the S-duality of Fibonacci.
 *
 *   - F_{n+1} = the "standard" execution capacity (forward momentum)
 *   - F_{n-1} = the "dual" execution capacity (inverse, low-friction)
 *   - L_n     = total capacity (standard + dual)
 *
 * When market volatility (coupling constant g) exceeds the crossover
 * threshold, the router flips from standard to dual execution. In the
 * dual space, extreme friction maps to frictionless math.
 *
 * The crossover is governed by the Cassini identity:
 *   F_{n+1}·F_{n-1} / F_n² = 1 ± 1/F_n²
 *
 * This means the product of standard and dual capacities is ALWAYS
 * approximately 1 (unit area preservation from SL(2,ℤ)). When standard
 * cost is high (g >> 1), dual cost is low (1/g << 1), and vice versa.
 *
 * Already proved in Lean:
 *   - lucas_gain: L_n = F_{n+1} + F_{n-1}
 *   - kam_cassini: F_{n-1}·F_{n+1} - F_n² = ±1
 *   - topological_gap: L_n = F_n + 2·F_{n-1}
 */

import {
  fib,
  lucas,
  fibonacciCassini,
  convergenceRatio,
  PHI,
  PHI_INV,
  PHI_SQUARED,
  SQRT5,
} from './boundary-topology';

// ============================================================================
// S-Duality Constants
// ============================================================================

/**
 * The S-duality crossover point.
 * When the coupling constant g equals this value, standard and dual
 * execution costs are equal. For the golden torus, this is g = 1
 * (because φ · 1/φ = 1).
 */
export const S_DUALITY_CROSSOVER = 1.0;

/**
 * The maximum stress bound.
 * Under S-duality, the effective stress is always min(g, 1/g) ≤ 1.
 * The system's maximum possible stress is capped at the crossover point.
 */
export const S_DUALITY_MAX_STRESS = 1.0;

// ============================================================================
// Execution Capacity
// ============================================================================

/** The execution capacity decomposition at dimension n. */
export interface SDualityCapacity {
  /** Dimension (Fibonacci index). */
  readonly dimension: number;
  /** Standard execution capacity: F_{n+1}. */
  readonly standardCapacity: number;
  /** Dual execution capacity: F_{n-1}. */
  readonly dualCapacity: number;
  /** Total capacity: L_n = standard + dual. */
  readonly totalCapacity: number;
  /** Standard/dual ratio: F_{n+1}/F_{n-1} → φ². */
  readonly capacityRatio: number;
  /** Cassini product: (F_{n+1}·F_{n-1}) / F_n² → 1 (area preservation). */
  readonly cassiniProduct: number;
  /** Is the S-duality decomposition valid? (total = L_n). */
  readonly valid: boolean;
}

/**
 * Decompose the execution capacity at dimension n into standard + dual.
 *
 * The fundamental S-duality identity: L_n = F_{n+1} + F_{n-1}.
 * Proved in Lean: lucas_gain.
 */
export function decomposeCapacity(n: number): SDualityCapacity {
  const standardCapacity = fib(n + 1);
  const dualCapacity = n > 0 ? fib(n - 1) : 0;
  const totalCapacity = lucas(n);
  const fn = fib(n);

  const capacityRatio = dualCapacity > 0 ? standardCapacity / dualCapacity : Infinity;
  const cassiniProduct = fn > 0
    ? (standardCapacity * dualCapacity) / (fn * fn)
    : 0;

  return {
    dimension: n,
    standardCapacity,
    dualCapacity,
    totalCapacity,
    capacityRatio,
    cassiniProduct,
    valid: standardCapacity + dualCapacity === totalCapacity,
  };
}

// ============================================================================
// S-Duality Router
// ============================================================================

/** The execution mode selected by the S-duality router. */
export type ExecutionMode = 'standard' | 'dual' | 'crossover';

/** Result of an S-duality routing decision. */
export interface SDualityRoutingDecision {
  /** The raw coupling constant (market volatility). */
  readonly couplingConstant: number;
  /** The effective stress after S-duality: min(g, 1/g). */
  readonly effectiveStress: number;
  /** Which execution mode was selected. */
  readonly mode: ExecutionMode;
  /** The capacity available in the selected mode. */
  readonly availableCapacity: number;
  /** The ratio reduction achieved by S-duality routing. */
  readonly stressReduction: number;
  /** The dimension used for capacity computation. */
  readonly dimension: number;
}

/**
 * Route an execution through S-duality.
 *
 * When the coupling constant g > 1, the router flips to dual execution
 * where the effective stress is 1/g. The Cassini identity guarantees
 * that standard × dual ≈ 1, so the total area is preserved.
 *
 * @param g - The coupling constant (market volatility / friction).
 * @param dimension - The Fibonacci dimension of the Swarm.
 */
export function routeSDuality(
  g: number,
  dimension: number
): SDualityRoutingDecision {
  const capacity = decomposeCapacity(dimension);
  const effectiveStress = Math.min(g, 1 / g);

  let mode: ExecutionMode;
  let availableCapacity: number;

  if (Math.abs(g - S_DUALITY_CROSSOVER) < 0.001) {
    // At the crossover: both paths are equally viable.
    mode = 'crossover';
    availableCapacity = capacity.totalCapacity;
  } else if (g < S_DUALITY_CROSSOVER) {
    // Low friction: use standard execution.
    mode = 'standard';
    availableCapacity = capacity.standardCapacity;
  } else {
    // High friction: flip to dual execution.
    mode = 'dual';
    availableCapacity = capacity.dualCapacity;
  }

  return {
    couplingConstant: g,
    effectiveStress,
    mode,
    availableCapacity,
    stressReduction: g > 0 ? 1 - effectiveStress / g : 0,
    dimension,
  };
}

/**
 * Compute the S-duality inversion of a coupling constant.
 *
 * g → 1/g
 *
 * In the Fibonacci lattice, this maps the forward ratio F_{n+1}/F_n → φ
 * to the backward ratio F_{n-1}/F_n → 1/φ.
 * Their product is always 1 ± 1/F_n² (Cassini).
 */
export function invertCoupling(g: number): number {
  if (g === 0) return Infinity;
  return 1 / g;
}

// ============================================================================
// The Golden S-Duality
// ============================================================================

/**
 * The Golden S-Duality identity.
 *
 * φ + 1/φ = √5
 *
 * The coupling constant (φ) plus its S-dual (1/φ) equals the
 * Hurwitz constant (√5). This means the total information capacity
 * of the S-dual Betti manifold is bounded by the universal
 * Diophantine resistance limit.
 *
 * Standard capacity (φ) + Dual capacity (1/φ) = Total capacity (√5)
 *
 * This is the continuous-limit version of L_n = F_{n+1} + F_{n-1}.
 */
export function verifyGoldenSDuality(): {
  readonly phi: number;
  readonly phiInverse: number;
  readonly sum: number;
  readonly sqrt5: number;
  readonly residual: number;
  readonly verified: boolean;
} {
  const sum = PHI + PHI_INV;
  const residual = Math.abs(sum - SQRT5);
  return {
    phi: PHI,
    phiInverse: PHI_INV,
    sum,
    sqrt5: SQRT5,
    residual,
    verified: residual < 1e-10,
  };
}

/**
 * Format a routing decision as a human-readable string.
 */
export function formatRoutingDecision(d: SDualityRoutingDecision): string {
  const reduction = (d.stressReduction * 100).toFixed(1);
  return [
    `[SDuality] g=${d.couplingConstant.toFixed(4)} → mode=${d.mode}`,
    `  effective stress: ${d.effectiveStress.toFixed(6)} (${reduction}% reduction)`,
    `  capacity: ${d.availableCapacity} (dim=${d.dimension})`,
  ].join('\n');
}
