// @a0n/.ignored_gnosis/src/betti/salem-vent-telemetry.ts
/**
 * salem-vent-telemetry.ts — Real-time KAM invariant monitoring.
 *
 * The Salem Vent Thread observes the frictionless orbits of the Betti
 * manifold in production. It monitors five invariants simultaneously:
 *
 *   1. Cassini determinant (±1) — SL(2,ℤ) armor integrity
 *   2. Pell discriminant (±4) — √5 convergence stability
 *   3. Reynolds number (L_n/F_n) — laminar flow health
 *   4. Diophantine resistance — armor thickness trending to 1/√5
 *   5. Phase-space conservation — Liouville volume preservation
 *
 * Emits structured telemetry events that can be consumed by dashboards,
 * alerting systems, or the Pisot immune system for automated response.
 */

import {
  type BoundaryTopology,
  type IsomorphicManifold,
  type BoundaryTransitionResult,
  createManifold,
  closeLoop,
  openRing,
  convergenceRatio,
  pellDiscriminant,
  fibonacciCassini,
  kamPhaseSpace,
  kamCoupling,
  diophantineResistance,
  kamSurvival,
  topologicalGap,
  PHI,
  PHI_SQUARED,
  SQRT5,
} from './boundary-topology';

// ============================================================================
// Telemetry Event Types
// ============================================================================

/** Severity levels for KAM telemetry events. */
export type KamSeverity = 'nominal' | 'warning' | 'critical' | 'breach';

/** A single KAM invariant observation. */
export interface KamInvariantObservation {
  /** Name of the invariant being observed. */
  readonly invariant: string;
  /** Current measured value. */
  readonly value: number;
  /** Expected value or range description. */
  readonly expected: string;
  /** Whether this invariant holds. */
  readonly intact: boolean;
}

/** A complete telemetry snapshot from the Salem Vent Thread. */
export interface SalemVentSnapshot {
  /** Monotonic sequence number. */
  readonly seq: number;
  /** Timestamp (ms since epoch). */
  readonly timestamp: number;
  /** The dimension being monitored. */
  readonly dimension: number;
  /** The current boundary condition. */
  readonly boundary: BoundaryTopology;

  // --- Core invariants ---
  /** Cassini determinant: should be ±1. */
  readonly cassiniDet: number;
  /** Pell discriminant residual: should be ±4. */
  readonly pellResidual: number;
  /** Reynolds number: L_n/F_n, converging to √5. */
  readonly reynoldsNumber: number;
  /** Diophantine resistance: |φ - F_{n+1}/F_n|·F_n², converging to 1/√5. */
  readonly armorThickness: number;
  /** Phase-space sum-of-squares identity check. */
  readonly phaseSpaceConserved: boolean;

  // --- Derived status ---
  /** Is the flow still laminar? (Reynolds < φ²). */
  readonly laminar: boolean;
  /** Is the SL(2,ℤ) armor intact? (Cassini = ±1). */
  readonly armorIntact: boolean;
  /** Is the Pell viscosity stable? (Pell = ±4). */
  readonly viscosityStable: boolean;
  /** Is the system fully immune per KAM theory? */
  readonly kamImmune: boolean;
  /** Overall severity assessment. */
  readonly severity: KamSeverity;

  /** Individual invariant observations. */
  readonly observations: readonly KamInvariantObservation[];
}

/** A boundary transition event detected by the telemetry. */
export interface SalemVentTransitionEvent {
  readonly type: 'boundary-transition';
  readonly timestamp: number;
  readonly from: BoundaryTopology;
  readonly to: BoundaryTopology;
  readonly dimension: number;
  readonly transition: BoundaryTransitionResult;
  /** Was the transition safe (laminar preserved)? */
  readonly safe: boolean;
}

/** A torus coupling event (two golden tori merging). */
export interface SalemVentCouplingEvent {
  readonly type: 'torus-coupling';
  readonly timestamp: number;
  readonly sourceM: number;
  readonly sourceN: number;
  readonly mergedIndex: number;
  readonly verified: boolean;
}

/** Any event emitted by the Salem Vent Thread. */
export type SalemVentEvent =
  | { readonly type: 'snapshot'; readonly data: SalemVentSnapshot }
  | SalemVentTransitionEvent
  | SalemVentCouplingEvent
  | { readonly type: 'armor-breach'; readonly data: SalemVentSnapshot };

/** Listener for Salem Vent events. */
export type SalemVentListener = (event: SalemVentEvent) => void;

// ============================================================================
// Salem Vent Thread
// ============================================================================

export interface SalemVentThreadOptions {
  /** Initial dimension to monitor (default: 10). */
  readonly initialDimension?: number;
  /** Initial boundary condition (default: 'Closed'). */
  readonly initialBoundary?: BoundaryTopology;
}

/**
 * The Salem Vent Thread.
 *
 * A stateful telemetry observer that monitors KAM invariants in real-time.
 * It maintains a current manifold state and emits structured events when:
 *   - An invariant is checked (snapshot)
 *   - A boundary transition occurs (Open ↔ Closed)
 *   - Two tori couple under load
 *   - An armor breach is detected
 *
 * The thread is designed to be driven by external ticks (from the Pisot
 * CPU's execution loop or the Swarm's heartbeat). It does not run its
 * own event loop — it is purely reactive.
 */
export class SalemVentThread {
  private seq = 0;
  private manifold: IsomorphicManifold;
  private readonly listeners: SalemVentListener[] = [];
  private history: SalemVentSnapshot[] = [];
  private readonly maxHistory: number;

  constructor(opts?: SalemVentThreadOptions) {
    const dim = opts?.initialDimension ?? 10;
    const boundary = opts?.initialBoundary ?? 'Closed';
    this.manifold = createManifold('salem-vent-live', dim, boundary);
    this.maxHistory = 1000;
  }

  /** Subscribe to telemetry events. Returns an unsubscribe function. */
  subscribe(listener: SalemVentListener): () => void {
    this.listeners.push(listener);
    return () => {
      const idx = this.listeners.indexOf(listener);
      if (idx >= 0) this.listeners.splice(idx, 1);
    };
  }

  private emit(event: SalemVentEvent): void {
    for (const listener of this.listeners) {
      try {
        listener(event);
      } catch {
        // Telemetry listeners must not crash the runtime.
      }
    }
  }

  /** Get the current manifold state. */
  getManifold(): IsomorphicManifold {
    return this.manifold;
  }

  /** Get the snapshot history (most recent last). */
  getHistory(): readonly SalemVentSnapshot[] {
    return this.history;
  }

  /**
   * Take a telemetry snapshot at the current dimension.
   * This is the primary tick function — call it on every Pisot CPU cycle
   * or Swarm heartbeat.
   */
  tick(): SalemVentSnapshot {
    const n = this.manifold.dimension;
    const cassiniDet = fibonacciCassini(n);
    const pellResidual = pellDiscriminant(n);
    const reynoldsNumber = convergenceRatio(n);
    const resistance = diophantineResistance(n);
    const phaseSpace = kamPhaseSpace(n);
    const survival = kamSurvival(n);

    const armorIntact = cassiniDet === 1 || cassiniDet === -1;
    const viscosityStable = pellResidual === 4 || pellResidual === -4;
    const laminar = reynoldsNumber < PHI_SQUARED;

    // Assess severity
    let severity: KamSeverity = 'nominal';
    if (!laminar) severity = 'warning';
    if (!armorIntact || !viscosityStable) severity = 'critical';
    if (!armorIntact && !viscosityStable) severity = 'breach';

    const observations: KamInvariantObservation[] = [
      {
        invariant: 'Cassini Determinant',
        value: cassiniDet,
        expected: '±1',
        intact: armorIntact,
      },
      {
        invariant: 'Pell Discriminant',
        value: pellResidual,
        expected: '±4',
        intact: viscosityStable,
      },
      {
        invariant: 'Reynolds Number',
        value: reynoldsNumber,
        expected: `< φ² ≈ ${PHI_SQUARED.toFixed(4)}`,
        intact: laminar,
      },
      {
        invariant: 'Diophantine Resistance',
        value: resistance.scaledError,
        expected: `→ 1/√5 ≈ ${(1 / SQRT5).toFixed(4)}`,
        intact: resistance.armorIntact,
      },
      {
        invariant: 'Phase-Space Conservation',
        value: phaseSpace.sumOfSquares,
        expected: `F_{${2 * n + 1}} = ${phaseSpace.doubledPlusOne}`,
        intact: phaseSpace.equal,
      },
    ];

    const snapshot: SalemVentSnapshot = {
      seq: this.seq++,
      timestamp: Date.now(),
      dimension: n,
      boundary: this.manifold.boundary,
      cassiniDet,
      pellResidual,
      reynoldsNumber,
      armorThickness: resistance.scaledError,
      phaseSpaceConserved: phaseSpace.equal,
      laminar,
      armorIntact,
      viscosityStable,
      kamImmune: survival.immune,
      severity,
      observations,
    };

    // Record history
    this.history.push(snapshot);
    if (this.history.length > this.maxHistory) {
      this.history = this.history.slice(-this.maxHistory);
    }

    // Emit events
    this.emit({ type: 'snapshot', data: snapshot });
    if (severity === 'breach') {
      this.emit({ type: 'armor-breach', data: snapshot });
    }

    return snapshot;
  }

  /**
   * Transition the manifold to a new boundary condition.
   * Emits a boundary-transition event with full safety analysis.
   */
  transitionBoundary(newBoundary: BoundaryTopology): SalemVentTransitionEvent {
    const oldBoundary = this.manifold.boundary;

    if (oldBoundary === newBoundary) {
      // No-op transition
      const result =
        newBoundary === 'Closed'
          ? closeLoop(this.manifold)
          : openRing(this.manifold);
      const event: SalemVentTransitionEvent = {
        type: 'boundary-transition',
        timestamp: Date.now(),
        from: oldBoundary,
        to: newBoundary,
        dimension: this.manifold.dimension,
        transition: result.transition,
        safe: true,
      };
      this.emit(event);
      return event;
    }

    const result =
      newBoundary === 'Closed'
        ? closeLoop(this.manifold)
        : openRing(this.manifold);

    this.manifold = result.manifold;

    const event: SalemVentTransitionEvent = {
      type: 'boundary-transition',
      timestamp: Date.now(),
      from: oldBoundary,
      to: newBoundary,
      dimension: this.manifold.dimension,
      transition: result.transition,
      safe: result.transition.laminar,
    };

    this.emit(event);
    return event;
  }

  /**
   * Scale the manifold to a new dimension.
   * Useful when the Swarm adds or removes edge workers.
   */
  scaleDimension(newDimension: number): void {
    this.manifold = createManifold(
      this.manifold.domainName,
      newDimension,
      this.manifold.boundary
    );
  }

  /**
   * Record a torus coupling event.
   * Called when two edge workers merge under load.
   */
  recordCoupling(m: number, n: number): SalemVentCouplingEvent {
    const coupling = kamCoupling(m, n);
    const event: SalemVentCouplingEvent = {
      type: 'torus-coupling',
      timestamp: Date.now(),
      sourceM: m,
      sourceN: n,
      mergedIndex: coupling.mergedIndex,
      verified: coupling.equal,
    };
    this.emit(event);
    return event;
  }

  /**
   * Format a snapshot as a human-readable string for logging.
   */
  static formatSnapshot(s: SalemVentSnapshot): string {
    const lines = [
      `[SalemVent #${s.seq}] dim=${s.dimension} boundary=${s.boundary} severity=${s.severity}`,
      `  Cassini: ${s.cassiniDet} (${s.armorIntact ? '✓' : '✗'})`,
      `  Pell:    ${s.pellResidual} (${s.viscosityStable ? '✓' : '✗'})`,
      `  Reynolds: ${s.reynoldsNumber.toFixed(6)} < φ²=${PHI_SQUARED.toFixed(4)} (${s.laminar ? 'laminar' : 'TURBULENT'})`,
      `  Armor:   ${s.armorThickness.toFixed(6)} → 1/√5=${(1 / SQRT5).toFixed(6)}`,
      `  Phase:   ${s.phaseSpaceConserved ? 'conserved' : 'VIOLATED'}`,
      `  KAM:     ${s.kamImmune ? 'IMMUNE' : 'VULNERABLE'}`,
    ];
    return lines.join('\n');
  }
}
