// @a0n/.ignored_gnosis/src/auth/auto-zk.ts
import type { ASTEdge, ASTNode, GraphAST } from '../betty/compiler';

export interface ZkAutoInjectionRecord {
  domain: 'sync' | 'materialization';
  sourceId: string;
  targetId: string;
  wrapperNodeId: string;
  reason: string;
}

export interface ZkAutoInjectionResult {
  ast: GraphAST;
  injected: ZkAutoInjectionRecord[];
}

const AUTO_INJECTED_EDGE_FLAG = 'zkAutoInjected';
const AUTO_INJECTED_NODE_FLAG = 'zkAutoInjected';
const AUTO_INJECTED_BY = 'zkAutoInjectedBy';
const AUTO_INJECTED_REASON = 'zkAutoReason';

function parseBoolean(value: string | undefined): boolean {
  if (!value) return false;
  const normalized = value.trim().toLowerCase();
  return normalized === 'true' || normalized === '1' || normalized === 'yes';
}

function hasCrossBoundarySync(properties: Record<string, string>): boolean {
  if (
    parseBoolean(properties.crossDevice) ||
    parseBoolean(properties.crossTenant)
  ) {
    return true;
  }

  const tenantScope = properties.tenantScope?.trim().toLowerCase();
  return tenantScope === 'cross-device' || tenantScope === 'cross-tenant';
}

function hasSensitiveMaterialization(
  properties: Record<string, string>
): boolean {
  const privateSignals =
    parseBoolean(properties.private) ||
    parseBoolean(properties.userPrivate) ||
    properties.visibility?.trim().toLowerCase() === 'private';

  if (!privateSignals) {
    return false;
  }

  const operation = properties.op?.trim().toLowerCase();
  const persistence = properties.persistence?.trim().toLowerCase();
  const storage = properties.storage?.trim().toLowerCase();

  return (
    operation === 'write_file' ||
    operation === 'read_file' ||
    persistence === 'fs.local' ||
    persistence === 'fs.durable' ||
    storage === 'fs.local' ||
    storage === 'fs.durable' ||
    storage === 'durable'
  );
}

function combineProperties(
  edge: ASTEdge,
  targetNode: ASTNode | undefined
): Record<string, string> {
  return {
    ...(targetNode?.properties ?? {}),
    ...edge.properties,
  };
}

function cloneNode(node: ASTNode): ASTNode {
  return {
    id: node.id,
    labels: [...node.labels],
    properties: { ...node.properties },
  };
}

function cloneEdge(edge: ASTEdge): ASTEdge {
  return {
    sourceIds: [...edge.sourceIds],
    targetIds: [...edge.targetIds],
    type: edge.type,
    properties: { ...edge.properties },
  };
}

function createUniqueWrapperNodeId(
  nodes: Map<string, ASTNode>,
  domain: 'sync' | 'materialization',
  sourceId: string,
  targetId: string,
  indexSeed: number
): string {
  const sourceSlug = sourceId.replace(/[^a-zA-Z0-9_]/g, '_');
  const targetSlug = targetId.replace(/[^a-zA-Z0-9_]/g, '_');
  let serial = indexSeed;

  while (true) {
    const candidate = `__zk_${domain}_${sourceSlug}_${targetSlug}_${serial}`;
    if (!nodes.has(candidate)) {
      return candidate;
    }
    serial += 1;
  }
}

function createWrapperNode(
  nodeId: string,
  domain: 'sync' | 'materialization',
  combinedProperties: Record<string, string>
): ASTNode {
  const label = domain === 'sync' ? 'ZKSyncEnvelope' : 'ZKMaterializeEnvelope';
  const properties: Record<string, string> = {
    zkMode: 'required',
    [AUTO_INJECTED_NODE_FLAG]: 'true',
    [AUTO_INJECTED_BY]: 'gnosis-auto-zk-injection',
    [AUTO_INJECTED_REASON]: `sensitive-${domain}`,
  };

  if (domain === 'sync') {
    if (combinedProperties.crossDevice) {
      properties.crossDevice = combinedProperties.crossDevice;
    }
    if (combinedProperties.crossTenant) {
      properties.crossTenant = combinedProperties.crossTenant;
    }
    if (combinedProperties.tenantScope) {
      properties.tenantScope = combinedProperties.tenantScope;
    }
  } else {
    if (combinedProperties.private) {
      properties.private = combinedProperties.private;
    } else {
      properties.private = 'true';
    }
    if (combinedProperties.visibility) {
      properties.visibility = combinedProperties.visibility;
    }
    if (combinedProperties.persistence) {
      properties.persistence = combinedProperties.persistence;
    }
    if (combinedProperties.storage) {
      properties.storage = combinedProperties.storage;
    }
    if (combinedProperties.op) {
      properties.op = combinedProperties.op;
    }
  }

  if (combinedProperties.recipientPublicKey) {
    properties.recipientPublicKey = combinedProperties.recipientPublicKey;
  }
  if (combinedProperties.zkRecipientPublicKey) {
    properties.zkRecipientPublicKey = combinedProperties.zkRecipientPublicKey;
  }
  if (combinedProperties.publicKey) {
    properties.publicKey = combinedProperties.publicKey;
  }

  return {
    id: nodeId,
    labels: [label],
    properties,
  };
}

function createInjectedEdge(
  sourceId: string,
  targetId: string,
  edge: ASTEdge
): ASTEdge {
  return {
    sourceIds: [sourceId],
    targetIds: [targetId],
    type: edge.type,
    properties: {
      ...edge.properties,
      [AUTO_INJECTED_EDGE_FLAG]: 'true',
    },
  };
}

function shouldSkipEdgeForInjection(edge: ASTEdge): boolean {
  if (edge.type !== 'PROCESS') {
    return true;
  }
  if (edge.sourceIds.length !== 1 || edge.targetIds.length !== 1) {
    return true;
  }
  if (parseBoolean(edge.properties[AUTO_INJECTED_EDGE_FLAG])) {
    return true;
  }
  return false;
}

/**
 * Handles the gnosis inject Sensitive Zk Envelopes workflow.
 */
export function injectSensitiveZkEnvelopes(
  ast: GraphAST
): ZkAutoInjectionResult {
  const nodes = new Map<string, ASTNode>();
  for (const [id, node] of ast.nodes.entries()) {
    nodes.set(id, cloneNode(node));
  }

  const rewrittenEdges: ASTEdge[] = [];
  const injected: ZkAutoInjectionRecord[] = [];
  let wrapperSeed = 0;

  for (const edge of ast.edges) {
    if (shouldSkipEdgeForInjection(edge)) {
      rewrittenEdges.push(cloneEdge(edge));
      continue;
    }

    const sourceId = (edge.sourceIds[0] ?? '').trim();
    const targetId = (edge.targetIds[0] ?? '').trim();
    const targetNode = nodes.get(targetId);
    const combined = combineProperties(edge, targetNode);
    const targetLabels = new Set(
      (targetNode?.labels ?? []).map((label) => label.trim().toLowerCase())
    );

    const needsSync =
      hasCrossBoundarySync(combined) && !targetLabels.has('zksyncenvelope');
    const needsMaterialization =
      hasSensitiveMaterialization(combined) &&
      !targetLabels.has('zkmaterializeenvelope');

    if (!needsSync && !needsMaterialization) {
      rewrittenEdges.push(cloneEdge(edge));
      continue;
    }

    let currentSourceId = sourceId;

    if (needsSync) {
      const wrapperNodeId = createUniqueWrapperNodeId(
        nodes,
        'sync',
        sourceId,
        targetId,
        wrapperSeed
      );
      wrapperSeed += 1;
      const wrapperNode = createWrapperNode(wrapperNodeId, 'sync', combined);
      nodes.set(wrapperNodeId, wrapperNode);
      rewrittenEdges.push(
        createInjectedEdge(currentSourceId, wrapperNodeId, edge)
      );
      injected.push({
        domain: 'sync',
        sourceId,
        targetId,
        wrapperNodeId,
        reason: 'cross-boundary sync properties detected',
      });
      currentSourceId = wrapperNodeId;
    }

    if (needsMaterialization) {
      const wrapperNodeId = createUniqueWrapperNodeId(
        nodes,
        'materialization',
        sourceId,
        targetId,
        wrapperSeed
      );
      wrapperSeed += 1;
      const wrapperNode = createWrapperNode(
        wrapperNodeId,
        'materialization',
        combined
      );
      nodes.set(wrapperNodeId, wrapperNode);
      rewrittenEdges.push(
        createInjectedEdge(currentSourceId, wrapperNodeId, edge)
      );
      injected.push({
        domain: 'materialization',
        sourceId,
        targetId,
        wrapperNodeId,
        reason: 'private materialization properties detected',
      });
      currentSourceId = wrapperNodeId;
    }

    rewrittenEdges.push(createInjectedEdge(currentSourceId, targetId, edge));
  }

  return {
    ast: {
      nodes,
      edges: rewrittenEdges,
    },
    injected,
  };
}

export const AUTO_ZK_EDGE_FLAG = AUTO_INJECTED_EDGE_FLAG;
export const AUTO_ZK_NODE_FLAG = AUTO_INJECTED_NODE_FLAG;
