// @a0n/.ignored_gnosis/src/auth/handlers.ts
import type {
  GnosisHandlerContext,
  GnosisRegistry,
} from '../runtime/registry';
import {
  checkGrantedCapability,
  delegateGranularUcan,
  generateUcanIdentity,
  type GnosisCapability,
  type GnosisEncryptedPayload,
  type GnosisIdentity,
  isAllowedCustodialAction,
  issueGranularUcan,
  verifyGranularUcan,
  zkDecryptUtf8,
  zkEncryptUtf8,
} from './core';
import {
  asHaltAttestationEnvelope,
  asHaltExecutionEnvelope,
  InMemoryNonceReplayStore,
  verifyHaltAttestation,
  verifyZkExecutionEnvelope,
  type HaltAttestationVerificationOptions,
  type ZkProofVerifier,
  type ZkProofVerifierInput,
} from './tee-attestation';
import {
  createEvmProofVerifier,
  type EvmProofVerifierConfig,
  type ZkProofEncoding,
} from './zk-onchain-verifier';

export type GnosisZkMode = 'required' | 'preferred' | 'off';
export type GnosisZkDomain =
  | 'delegation'
  | 'custodial'
  | 'sync'
  | 'materialization';

export interface GnosisZkPolicyReport {
  domain: GnosisZkDomain;
  mode: GnosisZkMode;
  sensitive: boolean;
  required: boolean;
  applied: boolean;
  status: 'applied' | 'already-encrypted' | 'skipped' | 'disabled';
  reason: string;
  sourceField?: string;
}

interface PayloadCandidate {
  field: string;
  plaintext: string;
}

interface ZkProtectionOptions {
  domain: GnosisZkDomain;
  mode: GnosisZkMode;
  sensitive: boolean;
  input: Record<string, unknown>;
  props: Record<string, string>;
  candidateFields: readonly string[];
}

interface ZkProtectionResult {
  encrypted: GnosisEncryptedPayload | null;
  report: GnosisZkPolicyReport;
}

const haltNonceReplayStore = new InMemoryNonceReplayStore();

function asRecord(value: unknown): Record<string, unknown> {
  return typeof value === 'object' && value !== null
    ? (value as Record<string, unknown>)
    : {};
}

function parseBoolean(value: string | undefined, fallback: boolean): boolean {
  if (!value) return fallback;
  const normalized = value.trim().toLowerCase();
  if (normalized === 'true') return true;
  if (normalized === 'false') return false;
  return fallback;
}

function readBoolean(value: unknown, fallback = false): boolean {
  if (typeof value === 'boolean') return value;
  if (typeof value === 'string') {
    const normalized = value.trim().toLowerCase();
    if (normalized === 'true') return true;
    if (normalized === 'false') return false;
  }
  return fallback;
}

function parseFiniteNumber(value: unknown): number | undefined {
  if (typeof value === 'number' && Number.isFinite(value)) {
    return value;
  }

  if (typeof value === 'string' && value.trim().length > 0) {
    const parsed = Number(value);
    if (Number.isFinite(parsed)) {
      return parsed;
    }
  }

  return undefined;
}

function parseStringList(value: unknown): string[] {
  if (Array.isArray(value)) {
    return value
      .filter((entry): entry is string => typeof entry === 'string')
      .map((entry) => entry.trim())
      .filter((entry) => entry.length > 0);
  }

  if (typeof value === 'string') {
    const raw = value.trim();
    if (raw.length === 0) {
      return [];
    }

    if (raw.startsWith('[')) {
      try {
        const parsed = JSON.parse(raw);
        return parseStringList(parsed);
      } catch {
        // Fall through to CSV parsing.
      }
    }

    return raw
      .split(',')
      .map((entry) => entry.trim())
      .filter((entry) => entry.length > 0);
  }

  return [];
}

function dedupeStrings(values: readonly string[]): string[] {
  return Array.from(new Set(values.map((value) => value.trim()))).filter(
    (value) => value.length > 0
  );
}

function parseRecordFromUnknown(
  value: unknown
): Record<string, unknown> | null {
  if (typeof value === 'string' && value.trim().length > 0) {
    try {
      const parsed = JSON.parse(value);
      return typeof parsed === 'object' &&
        parsed !== null &&
        !Array.isArray(parsed)
        ? (parsed as Record<string, unknown>)
        : null;
    } catch {
      return null;
    }
  }

  if (typeof value === 'object' && value !== null && !Array.isArray(value)) {
    return value as Record<string, unknown>;
  }

  return null;
}

function parseZkMode(
  value: string | undefined,
  fallback: GnosisZkMode
): GnosisZkMode {
  if (!value) return fallback;
  const normalized = value.trim().toLowerCase();
  if (normalized === 'required') return 'required';
  if (normalized === 'preferred') return 'preferred';
  if (normalized === 'off' || normalized === 'disabled') return 'off';
  return fallback;
}

function resolveZkMode(
  input: Record<string, unknown>,
  props: Record<string, string>,
  fallback: GnosisZkMode
): GnosisZkMode {
  const fromInput = typeof input.zkMode === 'string' ? input.zkMode : undefined;
  return parseZkMode(fromInput ?? props.zkMode, fallback);
}

function parseCapabilities(value: unknown): GnosisCapability[] {
  if (!Array.isArray(value)) {
    return [];
  }

  const capabilities: GnosisCapability[] = [];
  for (const entry of value) {
    const record = asRecord(entry);
    const can = record.can;
    const withValue = record.with;
    if (typeof can !== 'string' || typeof withValue !== 'string') {
      continue;
    }

    const capability: GnosisCapability = {
      can,
      with: withValue,
    };

    if (
      typeof record.constraints === 'object' &&
      record.constraints !== null &&
      !Array.isArray(record.constraints)
    ) {
      capability.constraints = record.constraints as Record<string, unknown>;
    }

    capabilities.push(capability);
  }

  return capabilities;
}

function parseCapabilitiesFromProps(
  props: Record<string, string>
): GnosisCapability[] {
  const raw = props.capabilities;
  if (!raw) {
    if (props.can && props.with) {
      return [{ can: props.can, with: props.with }];
    }
    return [];
  }

  try {
    const parsed = JSON.parse(raw);
    return parseCapabilities(parsed);
  } catch {
    if (props.can && props.with) {
      return [{ can: props.can, with: props.with }];
    }
    return [];
  }
}

function parseRequiredCapability(
  payload: Record<string, unknown>,
  props: Record<string, string>
): GnosisCapability | null {
  if (props.can && props.with) {
    return { can: props.can, with: props.with };
  }

  const candidate = asRecord(payload.requiredCapability);
  if (typeof candidate.can === 'string' && typeof candidate.with === 'string') {
    const constraints =
      typeof candidate.constraints === 'object' &&
      candidate.constraints !== null &&
      !Array.isArray(candidate.constraints)
        ? { constraints: candidate.constraints as Record<string, unknown> }
        : {};
    return {
      can: candidate.can,
      with: candidate.with,
      ...constraints,
    };
  }

  return null;
}

function parseEncryptedPayload(value: unknown): GnosisEncryptedPayload | null {
  const record = asRecord(value);
  if (
    typeof record.alg !== 'string' ||
    typeof record.ct !== 'string' ||
    typeof record.iv !== 'string' ||
    typeof record.tag !== 'string' ||
    typeof record.encryptedAt !== 'number'
  ) {
    return null;
  }

  return record as unknown as GnosisEncryptedPayload;
}

function parseJsonWebKey(value: unknown): JsonWebKey | undefined {
  if (typeof value === 'object' && value !== null && !Array.isArray(value)) {
    return value as JsonWebKey;
  }

  if (typeof value === 'string' && value.trim().length > 0) {
    try {
      const parsed = JSON.parse(value);
      if (
        typeof parsed === 'object' &&
        parsed !== null &&
        !Array.isArray(parsed)
      ) {
        return parsed as JsonWebKey;
      }
    } catch {
      return undefined;
    }
  }

  return undefined;
}

function parseJsonWebKeyList(value: unknown): JsonWebKey[] {
  if (Array.isArray(value)) {
    const keys: JsonWebKey[] = [];
    for (const candidate of value) {
      const key = parseJsonWebKey(candidate);
      if (key) {
        keys.push(key);
      }
    }
    return keys;
  }

  const key = parseJsonWebKey(value);
  return key ? [key] : [];
}

function parseTrustedKeyById(value: unknown): Record<string, JsonWebKey> {
  const record = parseRecordFromUnknown(value);
  if (!record) {
    return {};
  }

  const trustedKeys: Record<string, JsonWebKey> = {};
  for (const [keyId, candidate] of Object.entries(record)) {
    const parsed = parseJsonWebKey(candidate);
    if (parsed) {
      trustedKeys[keyId] = parsed;
    }
  }

  return trustedKeys;
}

function parseHaltVerificationOptions(
  input: Record<string, unknown>,
  props: Record<string, string>
): HaltAttestationVerificationOptions {
  const trustedMeasurements = dedupeStrings([
    ...parseStringList(input.trustedMeasurements),
    ...parseStringList(input.allowedMeasurements),
    ...parseStringList(props.trustedMeasurements),
    ...parseStringList(props.allowedMeasurements),
  ]);

  const trustedProgramHashes = dedupeStrings([
    ...parseStringList(input.trustedProgramHashes),
    ...parseStringList(input.allowedProgramHashes),
    ...parseStringList(props.trustedProgramHashes),
    ...parseStringList(props.allowedProgramHashes),
  ]);

  const trustedVkHashes = dedupeStrings([
    ...parseStringList(input.trustedVkHashes),
    ...parseStringList(input.allowedVkHashes),
    ...parseStringList(props.trustedVkHashes),
    ...parseStringList(props.allowedVkHashes),
  ]);

  const trustedPublicKeys = [
    ...parseJsonWebKeyList(input.trustedPublicKeys),
    ...parseJsonWebKeyList(input.trustedPublicKey),
    ...parseJsonWebKeyList(props.trustedPublicKeys),
    ...parseJsonWebKeyList(props.trustedPublicKey),
  ];

  const trustedKeyById = {
    ...parseTrustedKeyById(props.trustedKeyById),
    ...parseTrustedKeyById(input.trustedKeyById),
  };

  const maxAttestationAgeMs =
    parseFiniteNumber(input.maxAttestationAgeMs) ??
    parseFiniteNumber(props.maxAttestationAgeMs);

  const minTcbVersion =
    typeof input.minTcbVersion === 'string'
      ? input.minTcbVersion
      : props.minTcbVersion;

  const allowInlinePublicKey = readBoolean(
    input.allowInlinePublicKey,
    parseBoolean(props.allowInlinePublicKey, false)
  );

  const disableReplayProtection = readBoolean(
    input.disableReplayProtection,
    parseBoolean(props.disableReplayProtection, false)
  );

  return {
    ...(trustedMeasurements.length > 0 && { trustedMeasurements }),
    ...(trustedProgramHashes.length > 0 && { trustedProgramHashes }),
    ...(trustedVkHashes.length > 0 && { trustedVkHashes }),
    ...(trustedPublicKeys.length > 0 && { trustedPublicKeys }),
    ...(Object.keys(trustedKeyById).length > 0 && { trustedKeyById }),
    ...(maxAttestationAgeMs !== undefined && { maxAttestationAgeMs }),
    ...(minTcbVersion !== undefined && { minTcbVersion }),
    allowInlinePublicKey,
    ...(!disableReplayProtection && { nonceStore: haltNonceReplayStore }),
  };
}

function parseProofVerifier(value: unknown): ZkProofVerifier | undefined {
  if (typeof value !== 'function') {
    return undefined;
  }

  const verifier = value as (
    input: ZkProofVerifierInput
  ) => Promise<unknown> | unknown;

  return async (input: ZkProofVerifierInput) => {
    const result = await verifier(input);
    return result === true;
  };
}

function parseProofEncoding(value: unknown): ZkProofEncoding | undefined {
  if (typeof value !== 'string') {
    return undefined;
  }

  const normalized = value.trim().toLowerCase();
  if (normalized === 'auto') return 'auto';
  if (normalized === 'hex') return 'hex';
  if (normalized === 'base64') return 'base64';
  if (normalized === 'utf8') return 'utf8';
  return undefined;
}

function parseEvmProofVerifierConfig(
  input: Record<string, unknown>,
  props: Record<string, string>
): EvmProofVerifierConfig | null {
  const rpcUrl =
    typeof input.verifierRpcUrl === 'string'
      ? input.verifierRpcUrl
      : typeof input.rpcUrl === 'string'
      ? input.rpcUrl
      : props.verifierRpcUrl ?? props.rpcUrl;

  const verifierAddress =
    typeof input.verifierAddress === 'string'
      ? input.verifierAddress
      : props.verifierAddress;

  if (!rpcUrl || !verifierAddress) {
    return null;
  }

  const methodSelector =
    typeof input.verifierMethodSelector === 'string'
      ? input.verifierMethodSelector
      : props.verifierMethodSelector;

  const blockTag =
    typeof input.verifierBlockTag === 'string'
      ? input.verifierBlockTag
      : props.verifierBlockTag;

  const timeoutMs =
    parseFiniteNumber(input.verifierTimeoutMs) ??
    parseFiniteNumber(props.verifierTimeoutMs);

  const proofEncoding =
    parseProofEncoding(input.verifierProofEncoding) ??
    parseProofEncoding(props.verifierProofEncoding);

  return {
    rpcUrl,
    verifierAddress,
    ...(methodSelector !== undefined && { methodSelector }),
    ...(typeof blockTag === 'string' && blockTag.length > 0 && {
      blockTag: blockTag as NonNullable<EvmProofVerifierConfig['blockTag']>,
    }),
    ...(timeoutMs !== undefined && { timeoutMs }),
    ...(proofEncoding !== undefined && { proofEncoding }),
  };
}

function resolveProofVerifier(
  input: Record<string, unknown>,
  props: Record<string, string>
): ZkProofVerifier | undefined {
  const explicitVerifier = parseProofVerifier(input.proofVerifier);
  if (explicitVerifier) {
    return explicitVerifier;
  }

  const evmConfig = parseEvmProofVerifierConfig(input, props);
  if (evmConfig) {
    return createEvmProofVerifier(evmConfig);
  }

  return undefined;
}

function parseRecipientPublicKey(
  input: Record<string, unknown>,
  props: Record<string, string>
): JsonWebKey | undefined {
  const directCandidates = [
    input.recipientPublicKey,
    input.zkRecipientPublicKey,
    input.publicKey,
  ];
  for (const candidate of directCandidates) {
    const parsed = parseJsonWebKey(candidate);
    if (parsed) return parsed;
  }

  const propCandidates = [
    props.recipientPublicKey,
    props.zkRecipientPublicKey,
    props.publicKey,
  ];
  for (const candidate of propCandidates) {
    const parsed = parseJsonWebKey(candidate);
    if (parsed) return parsed;
  }

  return undefined;
}

function toPlaintext(value: unknown): string | null {
  if (typeof value === 'string') {
    return value.length > 0 ? value : null;
  }
  if (typeof value === 'number' || typeof value === 'boolean') {
    return String(value);
  }
  if (typeof value === 'bigint') {
    return value.toString(10);
  }
  if (typeof value === 'object' && value !== null) {
    try {
      return JSON.stringify(value);
    } catch {
      return null;
    }
  }
  return null;
}

function findPayloadCandidate(
  input: Record<string, unknown>,
  fields: readonly string[]
): PayloadCandidate | null {
  for (const field of fields) {
    if (!(field in input)) {
      continue;
    }
    const plaintext = toPlaintext(input[field]);
    if (plaintext !== null) {
      return { field, plaintext };
    }
  }
  return null;
}

function parseExistingEncryptedPayload(
  input: Record<string, unknown>
): GnosisEncryptedPayload | null {
  const candidates = [
    input.encrypted,
    input.encryptedPayload,
    input.zkEncrypted,
    input.zkEncryptedPayload,
  ];
  for (const candidate of candidates) {
    const parsed = parseEncryptedPayload(candidate);
    if (parsed) return parsed;
  }
  return null;
}

function hasDelegationConfidentialContext(
  input: Record<string, unknown>,
  props: Record<string, string>
): boolean {
  if (
    readBoolean(input.confidential, false) ||
    readBoolean(input.private, false) ||
    parseBoolean(props.confidential, false) ||
    parseBoolean(props.private, false)
  ) {
    return true;
  }

  const contextFields = ['delegationContext', 'context', 'attenuationContext'];
  return contextFields.some(
    (field) =>
      field in input &&
      input[field] !== null &&
      input[field] !== undefined &&
      input[field] !== ''
  );
}

function isCrossBoundarySync(
  input: Record<string, unknown>,
  props: Record<string, string>
): boolean {
  if (
    readBoolean(input.crossDevice, false) ||
    readBoolean(input.crossTenant, false) ||
    parseBoolean(props.crossDevice, false) ||
    parseBoolean(props.crossTenant, false)
  ) {
    return true;
  }

  const tenantScope =
    typeof input.tenantScope === 'string'
      ? input.tenantScope
      : props.tenantScope;
  return tenantScope === 'cross-tenant' || tenantScope === 'cross-device';
}

function isPrivateMaterialization(
  input: Record<string, unknown>,
  props: Record<string, string>
): boolean {
  if (
    readBoolean(input.private, false) ||
    readBoolean(input.userPrivate, false) ||
    parseBoolean(props.private, false) ||
    parseBoolean(props.userPrivate, false)
  ) {
    return true;
  }

  const visibility =
    typeof input.visibility === 'string' ? input.visibility : props.visibility;
  if (visibility === 'private') {
    return true;
  }

  const persistence =
    typeof input.persistence === 'string'
      ? input.persistence
      : props.persistence;
  return persistence === 'fs.local' || persistence === 'fs.durable';
}

async function protectPayloadWithZk(
  options: ZkProtectionOptions
): Promise<ZkProtectionResult> {
  const required = options.mode === 'required';
  const baseReport = {
    domain: options.domain,
    mode: options.mode,
    sensitive: options.sensitive,
    required,
  } as const;

  if (options.mode === 'off') {
    return {
      encrypted: null,
      report: {
        ...baseReport,
        applied: false,
        status: 'disabled',
        reason: 'ZK mode is disabled for this node.',
      },
    };
  }

  const existingEncrypted = parseExistingEncryptedPayload(options.input);
  if (existingEncrypted) {
    return {
      encrypted: existingEncrypted,
      report: {
        ...baseReport,
        applied: true,
        status: 'already-encrypted',
        reason: 'Encrypted payload already present.',
      },
    };
  }

  const candidate = findPayloadCandidate(
    options.input,
    options.candidateFields
  );
  if (!candidate) {
    return {
      encrypted: null,
      report: {
        ...baseReport,
        applied: false,
        status: 'skipped',
        reason: 'No payload field available for ZK protection.',
      },
    };
  }

  const recipientPublicKey = parseRecipientPublicKey(
    options.input,
    options.props
  );
  if (!recipientPublicKey) {
    if (required && options.sensitive) {
      throw new Error(
        `${options.domain} flow requires recipient public key when zkMode=required.`
      );
    }

    return {
      encrypted: null,
      report: {
        ...baseReport,
        applied: false,
        status: 'skipped',
        reason: 'Recipient public key not provided; encryption skipped.',
        sourceField: candidate.field,
      },
    };
  }

  const encrypted = await zkEncryptUtf8({
    plaintext: candidate.plaintext,
    recipientPublicKey,
    encryptionOptions: {
      category: options.domain,
    },
  });

  return {
    encrypted,
    report: {
      ...baseReport,
      applied: true,
      status: 'applied',
      reason: 'Sensitive payload encrypted with ZK policy.',
      sourceField: candidate.field,
    },
  };
}

export const GNOSIS_CORE_AUTH_LABELS = {
  UCAN_IDENTITY: 'UCANIdentity',
  UCAN_ISSUE: 'UCANIssue',
  UCAN_VERIFY: 'UCANVerify',
  UCAN_DELEGATE: 'UCANDelegate',
  UCAN_REQUIRE: 'UCANRequire',
  ZK_ENCRYPT: 'ZKEncrypt',
  ZK_DECRYPT: 'ZKDecrypt',
  CUSTODIAL_SIGNER: 'CustodialSigner',
  ZK_SYNC_ENVELOPE: 'ZKSyncEnvelope',
  ZK_MATERIALIZE_ENVELOPE: 'ZKMaterializeEnvelope',
  HALT_ATTESTATION_VERIFY: 'HALTAttestationVerify',
  ZK_EXECUTION_GATE: 'ZKExecutionGate',
} as const;

/**
 * Handles the gnosis register Core Auth Handlers workflow.
 */
export function registerCoreAuthHandlers(registry: GnosisRegistry): void {
  registry.register(
    GNOSIS_CORE_AUTH_LABELS.UCAN_IDENTITY,
    async (payload, props) => {
      const input = asRecord(payload);
      const displayName =
        typeof input.displayName === 'string'
          ? input.displayName
          : props.displayName;

      const identity = await generateUcanIdentity(
        displayName ? { displayName } : undefined
      );

      return {
        ...input,
        identity,
      };
    },
    { override: false }
  );

  registry.register(
    GNOSIS_CORE_AUTH_LABELS.UCAN_ISSUE,
    async (payload, props) => {
      const input = asRecord(payload);
      const issuer = input.identity as GnosisIdentity | undefined;
      const audience =
        typeof input.audience === 'string' ? input.audience : props.audience;

      if (!issuer) {
        throw new Error('UCANIssue requires payload.identity (Identity).');
      }
      if (!audience) {
        throw new Error(
          'UCANIssue requires an audience DID in payload.audience or props.audience.'
        );
      }

      const capabilities =
        parseCapabilities(input.capabilities) ||
        parseCapabilitiesFromProps(props);
      const effectiveCapabilities =
        capabilities.length > 0
          ? capabilities
          : parseCapabilitiesFromProps(props);

      if (effectiveCapabilities.length === 0) {
        throw new Error(
          'UCANIssue requires at least one capability (payload.capabilities or props.capabilities/can+with).'
        );
      }

      const expirationSeconds = Number.parseInt(
        props.expirationSeconds ?? `${input.expirationSeconds ?? ''}`,
        10
      );

      const token = await issueGranularUcan({
        issuer,
        audience,
        capabilities: effectiveCapabilities,
        ...(Number.isFinite(expirationSeconds) && {
          createOptions: { expirationSeconds },
        }),
      });

      return {
        ...input,
        ucan: {
          token,
          capabilities: effectiveCapabilities,
          audience,
        },
      };
    },
    { override: false }
  );

  registry.register(
    GNOSIS_CORE_AUTH_LABELS.UCAN_VERIFY,
    async (payload, props) => {
      const input = asRecord(payload);
      const token =
        typeof input.token === 'string'
          ? input.token
          : typeof asRecord(input.ucan).token === 'string'
          ? (asRecord(input.ucan).token as string)
          : null;
      const issuerPublicKey = input.issuerPublicKey as JsonWebKey | undefined;

      if (!token) {
        throw new Error(
          'UCANVerify requires payload.token or payload.ucan.token.'
        );
      }

      if (!issuerPublicKey) {
        throw new Error(
          'UCANVerify requires payload.issuerPublicKey (JsonWebKey).'
        );
      }

      const requiredCapabilities = parseCapabilitiesFromProps(props);
      const audience =
        typeof input.audience === 'string' ? input.audience : props.audience;

      const verification = await verifyGranularUcan({
        token,
        issuerPublicKey,
        verifyOptions: {
          ...(audience !== undefined && {
            audience: audience as `did:${string}:${string}`,
          }),
          ...(requiredCapabilities.length > 0 && { requiredCapabilities }),
        },
      });

      const failClosed = parseBoolean(props.failClosed, true);
      if (!verification.valid && failClosed) {
        throw new Error(
          `UCAN verification failed: ${verification.error ?? 'unknown error'}`
        );
      }

      const enforce = parseBoolean(props.enforce, true);
      const executionAuth = verification.valid
        ? {
            enforce,
            principal: verification.payload?.aud,
            token,
            capabilities: verification.payload?.att ?? [],
          }
        : undefined;

      return {
        ...input,
        ucanVerification: verification,
        executionAuth,
      };
    },
    { override: false }
  );

  registry.register(
    GNOSIS_CORE_AUTH_LABELS.UCAN_DELEGATE,
    async (payload, props) => {
      const input = asRecord(payload);
      const parentToken =
        typeof input.parentToken === 'string'
          ? input.parentToken
          : typeof input.token === 'string'
          ? input.token
          : null;
      const issuer = input.identity as GnosisIdentity | undefined;
      const audience =
        typeof input.audience === 'string' ? input.audience : props.audience;
      const capabilities =
        parseCapabilities(input.capabilities) ||
        parseCapabilitiesFromProps(props);
      const effectiveCapabilities =
        capabilities.length > 0
          ? capabilities
          : parseCapabilitiesFromProps(props);

      if (!parentToken) {
        throw new Error(
          'UCANDelegate requires payload.parentToken (or payload.token).'
        );
      }
      if (!issuer) {
        throw new Error('UCANDelegate requires payload.identity (Identity).');
      }
      if (!audience) {
        throw new Error(
          'UCANDelegate requires payload.audience or props.audience.'
        );
      }
      if (effectiveCapabilities.length === 0) {
        throw new Error('UCANDelegate requires delegated capabilities.');
      }

      const sensitiveDelegation = hasDelegationConfidentialContext(
        input,
        props
      );
      const delegateZkMode = resolveZkMode(
        input,
        props,
        sensitiveDelegation ? 'required' : 'preferred'
      );
      const delegationProtection = await protectPayloadWithZk({
        domain: 'delegation',
        mode: delegateZkMode,
        sensitive: sensitiveDelegation,
        input,
        props,
        candidateFields: ['delegationContext', 'context', 'attenuationContext'],
      });

      const delegatedToken = await delegateGranularUcan({
        parentToken,
        issuer,
        audience,
        capabilities: effectiveCapabilities,
        delegationOptions: {
          attenuate: parseBoolean(props.attenuate, true),
        },
      });

      return {
        ...input,
        delegatedUcan: {
          token: delegatedToken,
          capabilities: effectiveCapabilities,
          audience,
        },
        delegationContextEncrypted: delegationProtection.encrypted ?? undefined,
        zkPolicy: delegationProtection.report,
      };
    },
    { override: false }
  );

  registry.register(
    GNOSIS_CORE_AUTH_LABELS.UCAN_REQUIRE,
    async (payload, props, context?: GnosisHandlerContext) => {
      const input = asRecord(payload);
      const payloadExecutionAuth = asRecord(input.executionAuth);
      const contextExecutionAuth = asRecord(context?.executionAuth);
      const required = parseRequiredCapability(input, props);

      if (!required) {
        throw new Error(
          'UCANRequire requires can+with (props or payload.requiredCapability).'
        );
      }

      const capabilities = parseCapabilities(
        payloadExecutionAuth.capabilities ?? contextExecutionAuth.capabilities
      );
      const granted = checkGrantedCapability(capabilities, required);
      if (!granted) {
        throw new Error(
          `UCANRequire denied: missing ${required.can} on ${required.with}`
        );
      }

      return {
        ...input,
        executionAuth:
          input.executionAuth ?? context?.executionAuth ?? undefined,
        requiredCapability: required,
        requiredCapabilityGranted: true,
      };
    },
    { override: false }
  );

  registry.register(
    GNOSIS_CORE_AUTH_LABELS.ZK_ENCRYPT,
    async (payload, _props) => {
      const input = asRecord(payload);
      const plaintext =
        typeof input.plaintext === 'string'
          ? input.plaintext
          : typeof input.value === 'string'
          ? input.value
          : null;
      const recipientPublicKey = input.recipientPublicKey as
        | JsonWebKey
        | undefined;

      if (!plaintext) {
        throw new Error('ZKEncrypt requires payload.plaintext (string).');
      }
      if (!recipientPublicKey) {
        throw new Error(
          'ZKEncrypt requires payload.recipientPublicKey (JsonWebKey).'
        );
      }

      const encrypted = await zkEncryptUtf8({ plaintext, recipientPublicKey });

      return {
        ...input,
        encrypted,
      };
    },
    { override: false }
  );

  registry.register(
    GNOSIS_CORE_AUTH_LABELS.ZK_DECRYPT,
    async (payload, _props) => {
      const input = asRecord(payload);
      const encrypted = parseEncryptedPayload(input.encrypted);
      const recipientPrivateKey = input.recipientPrivateKey as
        | JsonWebKey
        | undefined;

      if (!encrypted) {
        throw new Error(
          'ZKDecrypt requires payload.encrypted (EncryptedPayload).'
        );
      }
      if (!recipientPrivateKey) {
        throw new Error(
          'ZKDecrypt requires payload.recipientPrivateKey (JsonWebKey).'
        );
      }

      const plaintext = await zkDecryptUtf8({
        encrypted,
        recipientPrivateKey,
      });

      return {
        ...input,
        plaintext,
      };
    },
    { override: false }
  );

  registry.register(
    GNOSIS_CORE_AUTH_LABELS.CUSTODIAL_SIGNER,
    async (payload, props) => {
      const input = asRecord(payload);
      const action =
        typeof input.action === 'string' ? input.action : props.action;

      if (!action) {
        throw new Error(
          'CustodialSigner requires an action (payload.action or props.action).'
        );
      }

      const allowed = isAllowedCustodialAction(action);
      if (!allowed && parseBoolean(props.failClosed, true)) {
        throw new Error(`CustodialSigner denied unknown action: ${action}`);
      }

      const hasCustodialPayload =
        findPayloadCandidate(input, [
          'payload',
          'signerPayload',
          'signature',
        ]) !== null;
      const custodialZkMode = resolveZkMode(
        input,
        props,
        hasCustodialPayload ? 'required' : 'preferred'
      );
      const custodialProtection = await protectPayloadWithZk({
        domain: 'custodial',
        mode: custodialZkMode,
        sensitive: hasCustodialPayload,
        input,
        props,
        candidateFields: ['payload', 'signerPayload', 'signature', 'request'],
      });

      return {
        ...input,
        encrypted: custodialProtection.encrypted ?? undefined,
        zkPolicy: custodialProtection.report,
        custodial: {
          action,
          allowed,
          payloadEncrypted: custodialProtection.report.applied,
        },
      };
    },
    { override: false }
  );

  registry.register(
    GNOSIS_CORE_AUTH_LABELS.ZK_SYNC_ENVELOPE,
    async (payload, props) => {
      const input = asRecord(payload);
      const syncSensitive = isCrossBoundarySync(input, props);
      const syncMode = resolveZkMode(
        input,
        props,
        syncSensitive ? 'required' : 'preferred'
      );
      const syncProtection = await protectPayloadWithZk({
        domain: 'sync',
        mode: syncMode,
        sensitive: syncSensitive,
        input,
        props,
        candidateFields: [
          'delta',
          'update',
          'payload',
          'state',
          'snapshot',
          'data',
        ],
      });

      return {
        ...input,
        encrypted: syncProtection.encrypted ?? undefined,
        zkPolicy: syncProtection.report,
        syncEnvelope: {
          sensitive: syncSensitive,
          encrypted: syncProtection.report.applied,
        },
      };
    },
    { override: false }
  );

  registry.register(
    GNOSIS_CORE_AUTH_LABELS.ZK_MATERIALIZE_ENVELOPE,
    async (payload, props) => {
      const input = asRecord(payload);
      const privateMaterialization = isPrivateMaterialization(input, props);
      const materializeMode = resolveZkMode(
        input,
        props,
        privateMaterialization ? 'required' : 'preferred'
      );
      const materializeProtection = await protectPayloadWithZk({
        domain: 'materialization',
        mode: materializeMode,
        sensitive: privateMaterialization,
        input,
        props,
        candidateFields: [
          'payload',
          'plaintext',
          'value',
          'state',
          'snapshot',
          'content',
        ],
      });

      return {
        ...input,
        encrypted: materializeProtection.encrypted ?? undefined,
        zkPolicy: materializeProtection.report,
        materializationEnvelope: {
          private: privateMaterialization,
          encrypted: materializeProtection.report.applied,
        },
      };
    },
    { override: false }
  );

  registry.register(
    GNOSIS_CORE_AUTH_LABELS.HALT_ATTESTATION_VERIFY,
    async (payload, props) => {
      const input = asRecord(payload);
      const attestation = asHaltAttestationEnvelope(input.attestation);
      if (!attestation) {
        throw new Error(
          'HALTAttestationVerify requires payload.attestation (HaltAttestationEnvelope).'
        );
      }

      const verification = await verifyHaltAttestation(
        attestation,
        parseHaltVerificationOptions(input, props)
      );
      const failClosed = readBoolean(
        input.failClosed,
        parseBoolean(props.failClosed, true)
      );

      if (!verification.valid && failClosed) {
        throw new Error(
          `HALTAttestationVerify failed: ${
            verification.reason ?? 'unknown failure'
          }`
        );
      }

      return {
        ...input,
        haltAttestation: verification,
      };
    },
    { override: false }
  );

  registry.register(
    GNOSIS_CORE_AUTH_LABELS.ZK_EXECUTION_GATE,
    async (payload, props) => {
      const input = asRecord(payload);
      const envelopeSource = Object.prototype.hasOwnProperty.call(
        input,
        'executionEnvelope'
      )
        ? input.executionEnvelope
        : input;
      const executionEnvelope = asHaltExecutionEnvelope(envelopeSource);
      if (!executionEnvelope) {
        throw new Error(
          'ZKExecutionGate requires payload.executionEnvelope with attestation/proof fields.'
        );
      }

      const proofVerifier = resolveProofVerifier(input, props);
      const proofVerifiedHint = readBoolean(input.proofVerified, false);
      const requireProofVerification = readBoolean(
        input.requireProofVerification,
        parseBoolean(props.requireProofVerification, true)
      );

      const gateResult = await verifyZkExecutionEnvelope(executionEnvelope, {
        ...parseHaltVerificationOptions(input, props),
        ...(proofVerifier !== undefined && { proofVerifier }),
        proofVerifiedHint,
        requireProofVerification,
      });
      const failClosed = readBoolean(
        input.failClosed,
        parseBoolean(props.failClosed, true)
      );

      if (!gateResult.valid && failClosed) {
        throw new Error(
          `ZKExecutionGate denied: ${gateResult.reason ?? 'unknown reason'}`
        );
      }

      return {
        ...input,
        executionEnvelope,
        executionGate: {
          allowed: gateResult.valid,
          reason: gateResult.reason,
          proofVerified: gateResult.proofVerified,
          attestation: gateResult.attestation,
        },
      };
    },
    { override: false }
  );
}
