// @a0n/.ignored_gnosis/src/auth/index.ts
export {
  authorizeSteeringApply,
  authorizeSteeringCapability,
  authorizeSteeringRelayConnect,
  authorizeSteeringRun,
  authorizeSteeringTraceAppend,
  authorizeTopologyEdge,
  buildEdgeResource,
  buildSteeringRelayResource,
  buildSteeringTopologyResource,
  buildSteeringTraceResource,
  checkGrantedCapability,
  mergeExecutionAuthContexts,
  normalizeExecutionAuthContext,
  topologyActionForEdge,
  generateUcanIdentity,
  issueGranularUcan,
  verifyGranularUcan,
  delegateGranularUcan,
  zkEncryptUtf8,
  zkDecryptUtf8,
  isAllowedCustodialAction,
} from './core';

export type {
  GnosisCapability,
  GnosisExecutionAuthContext,
  GnosisSteeringAuthorizationInput,
  GnosisSteeringAuthorizationResult,
  GnosisSteeringCapabilityAction,
  TopologyEdgeAuthorizationInput,
  TopologyEdgeAuthorizationResult,
} from './core';
export { injectSensitiveZkEnvelopes } from './auto-zk';
export type {
  ZkAutoInjectionRecord,
  ZkAutoInjectionResult,
} from './auto-zk';

export {
  GNOSIS_CORE_AUTH_LABELS,
  registerCoreAuthHandlers,
} from './handlers';
export type {
  GnosisZkDomain,
  GnosisZkMode,
  GnosisZkPolicyReport,
} from './handlers';

export {
  InMemoryNonceReplayStore,
  asHaltAttestationEnvelope,
  asHaltExecutionEnvelope,
  createHaltAttestation,
  hashPublicSignals,
  verifyHaltAttestation,
  verifyZkExecutionEnvelope,
} from './tee-attestation';
export type {
  HaltAttestationClaims,
  HaltAttestationEnvelope,
  HaltAttestationVerificationOptions,
  HaltAttestationVerificationResult,
  HaltExecutionEnvelope,
  NonceReplayStore,
  ZkExecutionVerificationOptions,
  ZkExecutionVerificationResult,
  ZkProofVerifier,
  ZkProofVerifierInput,
} from './tee-attestation';
export {
  createCertifiedRuntimeAttestation,
  createReactorHandleGrant,
  hashAeonReactorPolicy,
  hashCertifiedRuntimeBinding,
  verifyCertifiedRuntimeAttestation,
  verifyReactorHandleGrant,
} from './reactor-attestation';
export type {
  AeonReactorDefaultEffect,
  AeonReactorHandleKind,
  AeonReactorOperation,
  AeonReactorPolicy,
  AeonReactorPolicyRule,
  CertifiedRuntimeAttestationVerificationOptions,
  CertifiedRuntimeAttestationVerificationResult,
  CertifiedRuntimeBindingInput,
  CertifiedRuntimeClaims,
  CertifiedRuntimeEnvelope,
  CertifiedRuntimeSurface,
  CertifiedRuntimeTeeEvidence,
  CreateCertifiedRuntimeAttestationOptions,
  CreateReactorHandleGrantOptions,
  ReactorAttestationAlgorithm,
  ReactorHandleGrantClaims,
  ReactorHandleGrantEnvelope,
  ReactorHandleGrantVerificationOptions,
  ReactorHandleGrantVerificationResult,
  ReactorTeeProvider,
} from './reactor-attestation';
export { bootstrapExecutionAuthFromTopology } from './bootstrap';
export type {
  GnosisExecutionAuthBootstrapOptions,
  GnosisExecutionAuthBootstrapResult,
} from './bootstrap';

export {
  buildVerifyExecutionCalldata,
  createEvmProofVerifier,
  verifyProofViaEvmRpc,
} from './zk-onchain-verifier';
export type {
  EvmProofVerifierConfig,
  ZkProofEncoding,
} from './zk-onchain-verifier';
