// @a0n/.ignored_gnosis/src/auth/reactor-attestation.ts
import type { HostCapability } from '../capabilities/index';
import type { NonceReplayStore } from './tee-attestation';
import { decodeBase64Url, encodeBase64Url } from './encoding';

const REACTOR_ATTESTATION_ALGORITHM = 'ES256' as const;
const ECDSA_SIGNING_ALGORITHM: EcKeyImportParams = {
  name: 'ECDSA',
  namedCurve: 'P-256',
};
const ECDSA_SIGNATURE_ALGORITHM: EcdsaParams = {
  name: 'ECDSA',
  hash: 'SHA-256',
};

export type ReactorAttestationAlgorithm = typeof REACTOR_ATTESTATION_ALGORITHM;
export type AeonReactorHandleKind =
  | 'timer'
  | 'request'
  | 'response'
  | 'stream'
  | 'vfs'
  | 'task';
export type AeonReactorOperation =
  | 'timer.after'
  | 'timer.cancel'
  | 'request.read'
  | 'request.metadata'
  | 'response.start'
  | 'response.write'
  | 'response.end'
  | 'stream.open'
  | 'stream.read'
  | 'stream.write'
  | 'stream.close'
  | 'vfs.open'
  | 'vfs.read'
  | 'vfs.write'
  | 'vfs.delete'
  | 'task.spawn'
  | 'task.poll'
  | 'task.cancel';
export type AeonReactorDefaultEffect = 'allow' | 'deny';
export type ReactorTeeProvider = 'mock' | 'sgx' | 'sev-snp' | 'tdx' | 'nitro';
export type ReactorTeeEvidenceSource =
  | 'mock'
  | 'bridge'
  | 'host-agent'
  | 'host-sdk';
export type CertifiedRuntimeSurface =
  | 'http'
  | 'raw-flow'
  | 'gnode'
  | 'wasm'
  | 'vfs'
  | 'loopback-app';

export interface AeonReactorPolicyRule {
  readonly operation: AeonReactorOperation | '*';
  readonly effect?: AeonReactorDefaultEffect;
  readonly ucanCapability?: string;
  readonly hostCapability?: HostCapability;
  readonly resource?: string;
  readonly maxHandleTtlMs?: number;
  readonly notes?: string;
}

export interface AeonReactorPolicy {
  readonly version: 'aeon_reactor_v1';
  readonly runtime: string;
  readonly defaultEffect: AeonReactorDefaultEffect;
  readonly rules: readonly AeonReactorPolicyRule[];
}

export interface CertifiedRuntimeTeeEvidence {
  readonly source: ReactorTeeEvidenceSource;
  readonly provider: ReactorTeeProvider;
  readonly measurement: string;
  readonly reportDataHash: string;
  readonly evidence: {
    readonly format: string;
    readonly raw: string;
    readonly collateral?: Readonly<Record<string, string | readonly string[]>>;
  };
  readonly hostData?: string;
  readonly documentHash?: string;
  readonly tcbVersion?: string;
}

export interface CertifiedRuntimeBindingInput {
  readonly runtime: string;
  readonly audience: string;
  readonly subject?: string;
  readonly buildHash: string;
  readonly bundleDigest?: string;
  readonly policyHash: string;
  readonly lawDigest?: string;
  readonly hostCapabilities: readonly HostCapability[];
  readonly admittedOperations: readonly AeonReactorOperation[];
  readonly surfaces: readonly CertifiedRuntimeSurface[];
  readonly imageDigest?: string;
  readonly projectDigest?: string;
  readonly ucanIssuer?: string;
}

export interface CertifiedRuntimeClaims extends CertifiedRuntimeBindingInput {
  readonly bindingHash: string;
  readonly nonce: string;
  readonly issuedAt: number;
  readonly expiresAt: number;
  readonly tee: CertifiedRuntimeTeeEvidence;
}

export interface CertifiedRuntimeEnvelope {
  readonly algorithm: ReactorAttestationAlgorithm;
  readonly signature: string;
  readonly claims: CertifiedRuntimeClaims;
  readonly keyId?: string;
  readonly publicKey?: JsonWebKey;
}

export interface CreateCertifiedRuntimeAttestationOptions {
  readonly claims: Omit<
    CertifiedRuntimeClaims,
    'policyHash' | 'bindingHash' | 'tee'
  > & {
    readonly policy: AeonReactorPolicy;
    readonly tee: {
      readonly source?: ReactorTeeEvidenceSource;
      readonly provider: ReactorTeeProvider;
      readonly measurement: string;
      readonly evidence?: Partial<CertifiedRuntimeTeeEvidence['evidence']>;
      readonly hostData?: string;
      readonly documentHash?: string;
      readonly tcbVersion?: string;
    };
  };
  readonly privateKey: JsonWebKey;
  readonly keyId?: string;
  readonly publicKey?: JsonWebKey;
}

export interface CertifiedRuntimeAttestationVerificationOptions {
  readonly nowMs?: number;
  readonly maxAttestationAgeMs?: number;
  readonly trustedProviders?: readonly ReactorTeeProvider[];
  readonly trustedMeasurements?: readonly string[];
  readonly trustedPublicKeys?: readonly JsonWebKey[];
  readonly trustedKeyById?: Readonly<Record<string, JsonWebKey>>;
  readonly allowInlinePublicKey?: boolean;
  readonly minTcbVersion?: string;
  readonly expectedRuntime?: string;
  readonly expectedAudience?: string;
  readonly expectedSubject?: string;
  readonly expectedBuildHash?: string;
  readonly expectedBundleDigest?: string;
  readonly expectedPolicyHash?: string;
  readonly expectedPolicy?: AeonReactorPolicy;
  readonly expectedLawDigest?: string;
  readonly expectedImageDigest?: string;
  readonly expectedProjectDigest?: string;
  readonly expectedUcanIssuer?: string;
  readonly expectedBindingHash?: string;
  readonly expectedSurfaces?: readonly CertifiedRuntimeSurface[];
  readonly expectedAdmittedOperations?: readonly AeonReactorOperation[];
  readonly nonceStore?: NonceReplayStore;
}

export interface CertifiedRuntimeAttestationVerificationResult {
  readonly valid: boolean;
  readonly reason?: string;
  readonly claims?: CertifiedRuntimeClaims;
  readonly bindingHash?: string;
}

export interface ReactorHandleGrantClaims {
  readonly handleId: string;
  readonly kind: AeonReactorHandleKind;
  readonly ownerDid: string;
  readonly audience: string;
  readonly resource: string;
  readonly operations: readonly AeonReactorOperation[];
  readonly ucanCapabilities: readonly string[];
  readonly runtimeBindingHash: string;
  readonly delegationChainHash?: string;
  readonly issuedAt: number;
  readonly expiresAt: number;
}

export interface ReactorHandleGrantEnvelope {
  readonly algorithm: ReactorAttestationAlgorithm;
  readonly signature: string;
  readonly claims: ReactorHandleGrantClaims;
  readonly keyId?: string;
  readonly publicKey?: JsonWebKey;
}

export interface CreateReactorHandleGrantOptions {
  readonly claims: ReactorHandleGrantClaims;
  readonly privateKey: JsonWebKey;
  readonly keyId?: string;
  readonly publicKey?: JsonWebKey;
}

export interface ReactorHandleGrantVerificationOptions {
  readonly nowMs?: number;
  readonly trustedPublicKeys?: readonly JsonWebKey[];
  readonly trustedKeyById?: Readonly<Record<string, JsonWebKey>>;
  readonly allowInlinePublicKey?: boolean;
  readonly expectedHandleId?: string;
  readonly expectedOwnerDid?: string;
  readonly expectedAudience?: string;
  readonly expectedResource?: string;
  readonly expectedOperation?: AeonReactorOperation;
  readonly expectedUcanCapability?: string;
  readonly expectedRuntimeBindingHash?: string;
  readonly expectedDelegationChainHash?: string;
}

export interface ReactorHandleGrantVerificationResult {
  readonly valid: boolean;
  readonly reason?: string;
  readonly claims?: ReactorHandleGrantClaims;
}

function isNonEmptyString(value: unknown): value is string {
  return typeof value === 'string' && value.trim().length > 0;
}

function normalizeStringList(values: readonly string[]): string[] {
  return Array.from(
    new Set(
      values.map((value) => value.trim()).filter((value) => value.length > 0)
    )
  ).sort((left, right) => left.localeCompare(right));
}

function normalizeOptionalString(
  value: string | undefined
): string | undefined {
  return isNonEmptyString(value) ? value.trim() : undefined;
}

function normalizeOptionalCollateral(
  value: Readonly<Record<string, string | readonly string[]>> | undefined
): Readonly<Record<string, string | readonly string[]>> | undefined {
  if (!value) {
    return undefined;
  }

  const normalizedEntries = Object.entries(value)
    .map(([key, entry]) => {
      const normalizedKey = key.trim();
      if (normalizedKey.length === 0) {
        return null;
      }
      if (typeof entry === 'string') {
        const normalizedValue = entry.trim();
        return normalizedValue.length > 0
          ? ([normalizedKey, normalizedValue] as const)
          : null;
      }
      if (Array.isArray(entry)) {
        const normalizedValue = Array.from(
          new Set(
            entry
              .filter((item): item is string => typeof item === 'string')
              .map((item) => item.trim())
              .filter((item) => item.length > 0)
          )
        ).sort((left, right) => left.localeCompare(right));
        return normalizedValue.length > 0
          ? ([normalizedKey, normalizedValue] as const)
          : null;
      }
      return null;
    })
    .filter((entry): entry is NonNullable<typeof entry> => entry !== null)
    .sort(([left], [right]) => left.localeCompare(right));

  return normalizedEntries.length > 0
    ? Object.fromEntries(normalizedEntries)
    : undefined;
}

function quoteFormatForReactorTeeProvider(
  provider: ReactorTeeProvider
): string {
  switch (provider) {
    case 'mock':
      return 'mock-v1';
    case 'sgx':
      return 'sgx-quote';
    case 'sev-snp':
      return 'sev-snp-report';
    case 'tdx':
      return 'tdx-quote';
    case 'nitro':
      return 'nitro-document';
  }
}

function normalizeVersion(version: string): number[] {
  return version
    .split('.')
    .map((piece) => {
      const parsed = Number.parseInt(piece, 10);
      return Number.isFinite(parsed) ? parsed : 0;
    })
    .slice(0, 4);
}

function compareVersion(left: string, right: string): number {
  const leftParts = normalizeVersion(left);
  const rightParts = normalizeVersion(right);
  const maxLength = Math.max(leftParts.length, rightParts.length);

  for (let index = 0; index < maxLength; index += 1) {
    const leftValue = leftParts[index] ?? 0;
    const rightValue = rightParts[index] ?? 0;
    if (leftValue > rightValue) return 1;
    if (leftValue < rightValue) return -1;
  }

  return 0;
}

function stableStringify(value: unknown): string {
  if (value === null || value === undefined) {
    return 'null';
  }

  if (typeof value === 'string') {
    return JSON.stringify(value);
  }

  if (typeof value === 'number') {
    return Number.isFinite(value) ? String(value) : 'null';
  }

  if (typeof value === 'boolean') {
    return value ? 'true' : 'false';
  }

  if (typeof value === 'bigint') {
    return JSON.stringify(value.toString(10));
  }

  if (Array.isArray(value)) {
    return `[${value.map((item) => stableStringify(item)).join(',')}]`;
  }

  if (typeof value === 'object') {
    const record = value as Record<string, unknown>;
    const entries = Object.entries(record)
      .filter(([, entryValue]) => {
        const entryType = typeof entryValue;
        return (
          entryType !== 'undefined' &&
          entryType !== 'function' &&
          entryType !== 'symbol'
        );
      })
      .sort(([leftKey], [rightKey]) => leftKey.localeCompare(rightKey));

    return `{${entries
      .map(
        ([key, entryValue]) =>
          `${JSON.stringify(key)}:${stableStringify(entryValue)}`
      )
      .join(',')}}`;
  }

  return 'null';
}

function signaturePayload(claims: unknown): Uint8Array {
  return new TextEncoder().encode(stableStringify(claims));
}

function toArrayBuffer(bytes: Uint8Array): ArrayBuffer {
  return bytes.buffer.slice(
    bytes.byteOffset,
    bytes.byteOffset + bytes.byteLength
  ) as ArrayBuffer;
}

function toBase64Url(bytes: Uint8Array | ArrayBuffer): string {
  return encodeBase64Url(bytes);
}

function fromBase64Url(value: string): Uint8Array {
  return decodeBase64Url(value);
}

async function sha256Hex(input: string): Promise<string> {
  const digest = await crypto.subtle.digest(
    'SHA-256',
    new TextEncoder().encode(input)
  );
  return Array.from(new Uint8Array(digest), (byte) =>
    byte.toString(16).padStart(2, '0')
  ).join('');
}

async function importPublicKey(key: JsonWebKey): Promise<CryptoKey> {
  return crypto.subtle.importKey('jwk', key, ECDSA_SIGNING_ALGORITHM, false, [
    'verify',
  ]);
}

async function importPrivateKey(key: JsonWebKey): Promise<CryptoKey> {
  return crypto.subtle.importKey('jwk', key, ECDSA_SIGNING_ALGORITHM, false, [
    'sign',
  ]);
}

function resolveTrustedPublicKeys(
  envelope: CertifiedRuntimeEnvelope | ReactorHandleGrantEnvelope,
  options: {
    readonly trustedPublicKeys?: readonly JsonWebKey[];
    readonly trustedKeyById?: Readonly<Record<string, JsonWebKey>>;
    readonly allowInlinePublicKey?: boolean;
  }
): JsonWebKey[] {
  const resolved: JsonWebKey[] = [];

  if (envelope.keyId && options.trustedKeyById) {
    const key = options.trustedKeyById[envelope.keyId];
    if (key) {
      resolved.push(key);
    }
  }

  if (options.trustedPublicKeys && options.trustedPublicKeys.length > 0) {
    resolved.push(...options.trustedPublicKeys);
  }

  if (options.allowInlinePublicKey === true && envelope.publicKey) {
    resolved.push(envelope.publicKey);
  }

  const deduped = new Map<string, JsonWebKey>();
  for (const key of resolved) {
    deduped.set(stableStringify(key), key);
  }

  return Array.from(deduped.values());
}

async function verifyEnvelopeSignature(
  claims: unknown,
  signatureBase64Url: string,
  candidateKeys: readonly JsonWebKey[]
): Promise<boolean> {
  if (candidateKeys.length === 0) {
    return false;
  }

  const signature = fromBase64Url(signatureBase64Url);
  const payload = toArrayBuffer(signaturePayload(claims));
  const signatureBytes = toArrayBuffer(signature);

  for (const candidateKey of candidateKeys) {
    try {
      const key = await importPublicKey(candidateKey);
      const verified = await crypto.subtle.verify(
        ECDSA_SIGNATURE_ALGORITHM,
        key,
        signatureBytes,
        payload
      );
      if (verified) {
        return true;
      }
    } catch {
      // Continue checking additional trusted keys.
    }
  }

  return false;
}

async function signClaims(
  claims: unknown,
  privateKey: JsonWebKey
): Promise<string> {
  const key = await importPrivateKey(privateKey);
  const signature = await crypto.subtle.sign(
    ECDSA_SIGNATURE_ALGORITHM,
    key,
    toArrayBuffer(signaturePayload(claims))
  );
  return toBase64Url(signature);
}

function sameMembers(
  left: readonly string[],
  right: readonly string[]
): boolean {
  if (left.length !== right.length) {
    return false;
  }

  const normalizedLeft = normalizeStringList(left);
  const normalizedRight = normalizeStringList(right);
  return normalizedLeft.every(
    (value, index) => value === normalizedRight[index]
  );
}

function normalizePolicy(policy: AeonReactorPolicy): AeonReactorPolicy {
  return {
    version: 'aeon_reactor_v1',
    runtime: policy.runtime.trim(),
    defaultEffect: policy.defaultEffect,
    rules: policy.rules.map((rule) => {
      const ucanCapability = normalizeOptionalString(rule.ucanCapability);
      const resource = normalizeOptionalString(rule.resource);
      const notes = normalizeOptionalString(rule.notes);

      return {
        operation: rule.operation,
        ...(rule.effect ? { effect: rule.effect } : {}),
        ...(ucanCapability ? { ucanCapability } : {}),
        ...(rule.hostCapability ? { hostCapability: rule.hostCapability } : {}),
        ...(resource ? { resource } : {}),
        ...(typeof rule.maxHandleTtlMs === 'number' &&
        Number.isFinite(rule.maxHandleTtlMs)
          ? { maxHandleTtlMs: Math.max(0, Math.trunc(rule.maxHandleTtlMs)) }
          : {}),
        ...(notes ? { notes } : {}),
      };
    }),
  };
}

function normalizeBindingInput(
  input: CertifiedRuntimeBindingInput
): CertifiedRuntimeBindingInput {
  return {
    runtime: input.runtime.trim(),
    audience: input.audience.trim(),
    ...(normalizeOptionalString(input.subject)
      ? { subject: input.subject!.trim() }
      : {}),
    buildHash: input.buildHash.trim(),
    ...(normalizeOptionalString(input.bundleDigest)
      ? { bundleDigest: input.bundleDigest!.trim() }
      : {}),
    policyHash: input.policyHash.trim(),
    ...(normalizeOptionalString(input.lawDigest)
      ? { lawDigest: input.lawDigest!.trim() }
      : {}),
    hostCapabilities: normalizeStringList(
      input.hostCapabilities
    ) as HostCapability[],
    admittedOperations: normalizeStringList(
      input.admittedOperations
    ) as AeonReactorOperation[],
    surfaces: normalizeStringList(input.surfaces) as CertifiedRuntimeSurface[],
    ...(normalizeOptionalString(input.imageDigest)
      ? { imageDigest: input.imageDigest!.trim() }
      : {}),
    ...(normalizeOptionalString(input.projectDigest)
      ? { projectDigest: input.projectDigest!.trim() }
      : {}),
    ...(normalizeOptionalString(input.ucanIssuer)
      ? { ucanIssuer: input.ucanIssuer!.trim() }
      : {}),
  };
}

function normalizeHandleGrantClaims(
  claims: ReactorHandleGrantClaims
): ReactorHandleGrantClaims {
  return {
    handleId: claims.handleId.trim(),
    kind: claims.kind,
    ownerDid: claims.ownerDid.trim(),
    audience: claims.audience.trim(),
    resource: claims.resource.trim(),
    operations: normalizeStringList(
      claims.operations
    ) as AeonReactorOperation[],
    ucanCapabilities: normalizeStringList(claims.ucanCapabilities),
    runtimeBindingHash: claims.runtimeBindingHash.trim(),
    ...(normalizeOptionalString(claims.delegationChainHash)
      ? { delegationChainHash: claims.delegationChainHash!.trim() }
      : {}),
    issuedAt: claims.issuedAt,
    expiresAt: claims.expiresAt,
  };
}

function invalidRuntime(
  reason: string
): CertifiedRuntimeAttestationVerificationResult {
  return { valid: false, reason };
}

function invalidHandleGrant(
  reason: string
): ReactorHandleGrantVerificationResult {
  return { valid: false, reason };
}

function checkExpectedValue(
  actual: string | undefined,
  expected: string | undefined,
  label: string
): string | null {
  if (!expected) {
    return null;
  }

  if (actual !== expected) {
    return `${label} mismatch. Expected '${expected}', received '${
      actual ?? 'undefined'
    }'.`;
  }

  return null;
}

/**
 * Returns whether hash Aeon Reactor Policy is true.
 */
export async function hashAeonReactorPolicy(
  policy: AeonReactorPolicy
): Promise<string> {
  return sha256Hex(stableStringify(normalizePolicy(policy)));
}

/**
 * Returns whether hash Certified Runtime Binding is true.
 */
export async function hashCertifiedRuntimeBinding(
  input: CertifiedRuntimeBindingInput
): Promise<string> {
  return sha256Hex(stableStringify(normalizeBindingInput(input)));
}

/**
 * Creates the Certified Runtime Attestation.
 */
export async function createCertifiedRuntimeAttestation(
  options: CreateCertifiedRuntimeAttestationOptions
): Promise<CertifiedRuntimeEnvelope> {
  const policyHash = await hashAeonReactorPolicy(options.claims.policy);
  const teeEvidence = options.claims.tee.evidence ?? {};
  const normalizedEvidence =
    teeEvidence && typeof teeEvidence === 'object'
      ? teeEvidence
      : ({} as Partial<CertifiedRuntimeTeeEvidence['evidence']>);
  const bindingInput = normalizeBindingInput({
    runtime: options.claims.runtime,
    audience: options.claims.audience,
    ...(options.claims.subject ? { subject: options.claims.subject } : {}),
    buildHash: options.claims.buildHash,
    ...(options.claims.bundleDigest
      ? { bundleDigest: options.claims.bundleDigest }
      : {}),
    policyHash,
    ...(options.claims.lawDigest
      ? { lawDigest: options.claims.lawDigest }
      : {}),
    hostCapabilities: options.claims.hostCapabilities,
    admittedOperations: options.claims.admittedOperations,
    surfaces: options.claims.surfaces,
    ...(options.claims.imageDigest
      ? { imageDigest: options.claims.imageDigest }
      : {}),
    ...(options.claims.projectDigest
      ? { projectDigest: options.claims.projectDigest }
      : {}),
    ...(options.claims.ucanIssuer
      ? { ucanIssuer: options.claims.ucanIssuer }
      : {}),
  });
  const bindingHash = await hashCertifiedRuntimeBinding(bindingInput);
  const claims: CertifiedRuntimeClaims = {
    ...bindingInput,
    bindingHash,
    nonce: options.claims.nonce.trim(),
    issuedAt: options.claims.issuedAt,
    expiresAt: options.claims.expiresAt,
    tee: {
      source: options.claims.tee.source ?? 'mock',
      provider: options.claims.tee.provider,
      measurement: options.claims.tee.measurement.trim(),
      reportDataHash: bindingHash,
      evidence: {
        format:
          normalizeOptionalString(normalizedEvidence.format) ??
          quoteFormatForReactorTeeProvider(options.claims.tee.provider),
        raw:
          normalizeOptionalString(normalizedEvidence.raw) ??
          options.claims.tee.measurement.trim(),
        ...(normalizeOptionalCollateral(normalizedEvidence.collateral)
          ? {
              collateral: normalizeOptionalCollateral(
                normalizedEvidence.collateral
              )!,
            }
          : {}),
      },
      ...(normalizeOptionalString(options.claims.tee.hostData)
        ? { hostData: options.claims.tee.hostData!.trim() }
        : {}),
      ...(normalizeOptionalString(options.claims.tee.documentHash)
        ? { documentHash: options.claims.tee.documentHash!.trim() }
        : {}),
      ...(normalizeOptionalString(options.claims.tee.tcbVersion)
        ? { tcbVersion: options.claims.tee.tcbVersion!.trim() }
        : {}),
    },
  };

  return {
    algorithm: REACTOR_ATTESTATION_ALGORITHM,
    signature: await signClaims(claims, options.privateKey),
    claims,
    ...(options.keyId ? { keyId: options.keyId } : {}),
    ...(options.publicKey ? { publicKey: options.publicKey } : {}),
  };
}

/**
 * Handles the gnosis verify Certified Runtime Attestation workflow.
 */
export async function verifyCertifiedRuntimeAttestation(
  envelope: CertifiedRuntimeEnvelope,
  options: CertifiedRuntimeAttestationVerificationOptions = {}
): Promise<CertifiedRuntimeAttestationVerificationResult> {
  const nowMs = options.nowMs ?? Date.now();
  const trustedPublicKeys = resolveTrustedPublicKeys(envelope, options);
  if (trustedPublicKeys.length === 0) {
    return invalidRuntime(
      'No trusted public keys resolved. Provide trustedKeyById/trustedPublicKeys or allow inline key explicitly.'
    );
  }

  if (envelope.algorithm !== REACTOR_ATTESTATION_ALGORITHM) {
    return invalidRuntime(
      `Unsupported reactor attestation algorithm '${envelope.algorithm}'.`
    );
  }

  const signatureValid = await verifyEnvelopeSignature(
    envelope.claims,
    envelope.signature,
    trustedPublicKeys
  );
  if (!signatureValid) {
    return invalidRuntime('Runtime attestation signature verification failed.');
  }

  const claims = envelope.claims;
  if (!isNonEmptyString(claims.runtime)) {
    return invalidRuntime(
      'Runtime attestation runtime must be a non-empty string.'
    );
  }
  if (!isNonEmptyString(claims.audience)) {
    return invalidRuntime(
      'Runtime attestation audience must be a non-empty string.'
    );
  }
  if (!isNonEmptyString(claims.buildHash)) {
    return invalidRuntime(
      'Runtime attestation buildHash must be a non-empty string.'
    );
  }
  if (!isNonEmptyString(claims.policyHash)) {
    return invalidRuntime(
      'Runtime attestation policyHash must be a non-empty string.'
    );
  }
  if (!isNonEmptyString(claims.bindingHash)) {
    return invalidRuntime(
      'Runtime attestation bindingHash must be a non-empty string.'
    );
  }
  if (!isNonEmptyString(claims.nonce)) {
    return invalidRuntime(
      'Runtime attestation nonce must be a non-empty string.'
    );
  }
  if (
    !isNonEmptyString(claims.tee.source) ||
    !isNonEmptyString(claims.tee.measurement) ||
    !isNonEmptyString(claims.tee.reportDataHash) ||
    !claims.tee.evidence ||
    !isNonEmptyString(claims.tee.evidence.format) ||
    !isNonEmptyString(claims.tee.evidence.raw)
  ) {
    return invalidRuntime(
      'Runtime attestation TEE evidence must include source, measurement, reportDataHash, and evidence payload.'
    );
  }

  if (!Number.isFinite(claims.issuedAt) || !Number.isFinite(claims.expiresAt)) {
    return invalidRuntime(
      'Runtime attestation timestamps must be finite epoch milliseconds.'
    );
  }
  if (claims.issuedAt > nowMs) {
    return invalidRuntime('Runtime attestation issuedAt is in the future.');
  }
  if (claims.expiresAt <= claims.issuedAt) {
    return invalidRuntime(
      'Runtime attestation expiresAt must be greater than issuedAt.'
    );
  }
  if (claims.expiresAt <= nowMs) {
    return invalidRuntime('Runtime attestation is expired.');
  }
  if (
    typeof options.maxAttestationAgeMs === 'number' &&
    options.maxAttestationAgeMs >= 0 &&
    nowMs - claims.issuedAt > options.maxAttestationAgeMs
  ) {
    return invalidRuntime(
      `Runtime attestation exceeds max age of ${options.maxAttestationAgeMs} ms.`
    );
  }

  if (
    options.trustedProviders &&
    options.trustedProviders.length > 0 &&
    !options.trustedProviders.includes(claims.tee.provider)
  ) {
    return invalidRuntime(
      `TEE provider '${claims.tee.provider}' is not trusted.`
    );
  }

  if (
    options.trustedMeasurements &&
    options.trustedMeasurements.length > 0 &&
    !options.trustedMeasurements.includes(claims.tee.measurement)
  ) {
    return invalidRuntime(
      `TEE measurement '${claims.tee.measurement}' is not trusted.`
    );
  }

  if (options.minTcbVersion) {
    if (!claims.tee.tcbVersion) {
      return invalidRuntime(
        `TEE evidence is missing tcbVersion; minimum '${options.minTcbVersion}' required.`
      );
    }
    if (compareVersion(claims.tee.tcbVersion, options.minTcbVersion) < 0) {
      return invalidRuntime(
        `TEE tcbVersion '${claims.tee.tcbVersion}' is below required '${options.minTcbVersion}'.`
      );
    }
  }

  const expectedPolicyHash =
    options.expectedPolicyHash ??
    (options.expectedPolicy
      ? await hashAeonReactorPolicy(options.expectedPolicy)
      : undefined);
  const expectedMismatches = [
    checkExpectedValue(claims.runtime, options.expectedRuntime, 'runtime'),
    checkExpectedValue(claims.audience, options.expectedAudience, 'audience'),
    checkExpectedValue(claims.subject, options.expectedSubject, 'subject'),
    checkExpectedValue(
      claims.buildHash,
      options.expectedBuildHash,
      'buildHash'
    ),
    checkExpectedValue(
      claims.bundleDigest,
      options.expectedBundleDigest,
      'bundleDigest'
    ),
    checkExpectedValue(claims.policyHash, expectedPolicyHash, 'policyHash'),
    checkExpectedValue(
      claims.lawDigest,
      options.expectedLawDigest,
      'lawDigest'
    ),
    checkExpectedValue(
      claims.imageDigest,
      options.expectedImageDigest,
      'imageDigest'
    ),
    checkExpectedValue(
      claims.projectDigest,
      options.expectedProjectDigest,
      'projectDigest'
    ),
    checkExpectedValue(
      claims.ucanIssuer,
      options.expectedUcanIssuer,
      'ucanIssuer'
    ),
    checkExpectedValue(
      claims.bindingHash,
      options.expectedBindingHash,
      'bindingHash'
    ),
  ].find((issue) => issue !== null);
  if (expectedMismatches) {
    return invalidRuntime(expectedMismatches);
  }

  const recomputedBindingHash = await hashCertifiedRuntimeBinding({
    runtime: claims.runtime,
    audience: claims.audience,
    ...(claims.subject ? { subject: claims.subject } : {}),
    buildHash: claims.buildHash,
    ...(claims.bundleDigest ? { bundleDigest: claims.bundleDigest } : {}),
    policyHash: claims.policyHash,
    ...(claims.lawDigest ? { lawDigest: claims.lawDigest } : {}),
    hostCapabilities: claims.hostCapabilities,
    admittedOperations: claims.admittedOperations,
    surfaces: claims.surfaces,
    ...(claims.imageDigest ? { imageDigest: claims.imageDigest } : {}),
    ...(claims.projectDigest ? { projectDigest: claims.projectDigest } : {}),
    ...(claims.ucanIssuer ? { ucanIssuer: claims.ucanIssuer } : {}),
  });
  if (recomputedBindingHash !== claims.bindingHash) {
    return invalidRuntime(
      `Runtime bindingHash mismatch. Expected '${recomputedBindingHash}', received '${claims.bindingHash}'.`
    );
  }
  if (claims.tee.reportDataHash !== claims.bindingHash) {
    return invalidRuntime(
      `TEE reportDataHash '${claims.tee.reportDataHash}' does not match runtime bindingHash '${claims.bindingHash}'.`
    );
  }

  if (
    options.expectedSurfaces &&
    !sameMembers(claims.surfaces, options.expectedSurfaces)
  ) {
    return invalidRuntime(
      'Runtime surfaces do not match the expected certified surface set.'
    );
  }
  if (
    options.expectedAdmittedOperations &&
    !sameMembers(claims.admittedOperations, options.expectedAdmittedOperations)
  ) {
    return invalidRuntime(
      'Admitted reactor operations do not match the expected operation set.'
    );
  }

  if (options.nonceStore) {
    options.nonceStore.prune(nowMs);
    if (options.nonceStore.has(claims.nonce, nowMs)) {
      return invalidRuntime(
        `Nonce replay detected for runtime attestation '${claims.nonce}'.`
      );
    }
    options.nonceStore.mark(claims.nonce, claims.expiresAt);
  }

  return {
    valid: true,
    claims,
    bindingHash: claims.bindingHash,
  };
}

/**
 * Creates the Reactor Handle Grant.
 */
export async function createReactorHandleGrant(
  options: CreateReactorHandleGrantOptions
): Promise<ReactorHandleGrantEnvelope> {
  const claims = normalizeHandleGrantClaims(options.claims);

  return {
    algorithm: REACTOR_ATTESTATION_ALGORITHM,
    signature: await signClaims(claims, options.privateKey),
    claims,
    ...(options.keyId ? { keyId: options.keyId } : {}),
    ...(options.publicKey ? { publicKey: options.publicKey } : {}),
  };
}

/**
 * Handles the gnosis verify Reactor Handle Grant workflow.
 */
export async function verifyReactorHandleGrant(
  envelope: ReactorHandleGrantEnvelope,
  options: ReactorHandleGrantVerificationOptions = {}
): Promise<ReactorHandleGrantVerificationResult> {
  const nowMs = options.nowMs ?? Date.now();
  const trustedPublicKeys = resolveTrustedPublicKeys(envelope, options);
  if (trustedPublicKeys.length === 0) {
    return invalidHandleGrant(
      'No trusted public keys resolved. Provide trustedKeyById/trustedPublicKeys or allow inline key explicitly.'
    );
  }

  if (envelope.algorithm !== REACTOR_ATTESTATION_ALGORITHM) {
    return invalidHandleGrant(
      `Unsupported reactor handle grant algorithm '${envelope.algorithm}'.`
    );
  }

  const signatureValid = await verifyEnvelopeSignature(
    envelope.claims,
    envelope.signature,
    trustedPublicKeys
  );
  if (!signatureValid) {
    return invalidHandleGrant(
      'Reactor handle grant signature verification failed.'
    );
  }

  const claims = envelope.claims;
  if (
    !isNonEmptyString(claims.handleId) ||
    !isNonEmptyString(claims.ownerDid) ||
    !isNonEmptyString(claims.audience) ||
    !isNonEmptyString(claims.resource) ||
    !isNonEmptyString(claims.runtimeBindingHash)
  ) {
    return invalidHandleGrant(
      'Reactor handle grant requires non-empty handleId, ownerDid, audience, resource, and runtimeBindingHash.'
    );
  }
  if (!Number.isFinite(claims.issuedAt) || !Number.isFinite(claims.expiresAt)) {
    return invalidHandleGrant(
      'Reactor handle grant timestamps must be finite epoch milliseconds.'
    );
  }
  if (claims.issuedAt > nowMs) {
    return invalidHandleGrant(
      'Reactor handle grant issuedAt is in the future.'
    );
  }
  if (claims.expiresAt <= claims.issuedAt) {
    return invalidHandleGrant(
      'Reactor handle grant expiresAt must be greater than issuedAt.'
    );
  }
  if (claims.expiresAt <= nowMs) {
    return invalidHandleGrant('Reactor handle grant is expired.');
  }

  const expectedMismatches = [
    checkExpectedValue(claims.handleId, options.expectedHandleId, 'handleId'),
    checkExpectedValue(claims.ownerDid, options.expectedOwnerDid, 'ownerDid'),
    checkExpectedValue(claims.audience, options.expectedAudience, 'audience'),
    checkExpectedValue(claims.resource, options.expectedResource, 'resource'),
    checkExpectedValue(
      claims.runtimeBindingHash,
      options.expectedRuntimeBindingHash,
      'runtimeBindingHash'
    ),
    checkExpectedValue(
      claims.delegationChainHash,
      options.expectedDelegationChainHash,
      'delegationChainHash'
    ),
  ].find((issue) => issue !== null);
  if (expectedMismatches) {
    return invalidHandleGrant(expectedMismatches);
  }

  if (
    options.expectedOperation &&
    !claims.operations.includes(options.expectedOperation)
  ) {
    return invalidHandleGrant(
      `Expected operation '${options.expectedOperation}' is not admitted by the handle grant.`
    );
  }
  if (
    options.expectedUcanCapability &&
    !claims.ucanCapabilities.includes(options.expectedUcanCapability)
  ) {
    return invalidHandleGrant(
      `Expected UCAN capability '${options.expectedUcanCapability}' is not admitted by the handle grant.`
    );
  }

  return {
    valid: true,
    claims,
  };
}
