// @a0n/.ignored_gnosis/src/auth/zk-onchain-verifier.ts
import type {
  ZkProofVerifier,
  ZkProofVerifierInput,
} from './tee-attestation';
import { decodeBase64, decodeHex, encodeHex } from './encoding';

export type ZkProofEncoding = 'auto' | 'hex' | 'base64' | 'utf8';

export interface EvmProofVerifierConfig {
  rpcUrl: string;
  verifierAddress: string;
  methodSelector?: string;
  blockTag?: 'latest' | 'pending' | `0x${string}`;
  timeoutMs?: number;
  proofEncoding?: ZkProofEncoding;
}

const DEFAULT_BLOCK_TAG = 'latest' as const;
const DEFAULT_TIMEOUT_MS = 10_000;
const DEFAULT_VERIFY_METHOD_SELECTOR = '0x4d48f5fe';

function normalizeHex(value: string): string {
  const trimmed = value.trim();
  return trimmed.startsWith('0x') || trimmed.startsWith('0X')
    ? trimmed.slice(2)
    : trimmed;
}

function isHexString(value: string): boolean {
  return /^[0-9a-fA-F]+$/.test(value);
}

function padHexLeft(hexValue: string, targetLength: number): string {
  return hexValue.padStart(targetLength, '0');
}

function padHexRight(hexValue: string, targetLength: number): string {
  return hexValue.padEnd(targetLength, '0');
}

function encodeWord(hexValue: string): string {
  const normalized = normalizeHex(hexValue);
  if (!isHexString(normalized)) {
    throw new Error('Invalid hex value for ABI word encoding.');
  }
  if (normalized.length > 64) {
    throw new Error('ABI word encoding overflow: value exceeds 32 bytes.');
  }
  return padHexLeft(normalized, 64);
}

function uintToHex(value: number): string {
  if (!Number.isInteger(value) || value < 0) {
    throw new Error('Expected a non-negative integer for ABI uint encoding.');
  }
  return value.toString(16);
}

function normalizeMethodSelector(value: string | undefined): string {
  const candidate = value ?? DEFAULT_VERIFY_METHOD_SELECTOR;
  const normalized = normalizeHex(candidate).toLowerCase();
  if (!isHexString(normalized) || normalized.length !== 8) {
    throw new Error(
      'Verifier method selector must be exactly 4 bytes (8 hex chars).'
    );
  }
  return `0x${normalized}`;
}

function normalizeAddress(value: string): string {
  const normalized = normalizeHex(value).toLowerCase();
  if (!/^[0-9a-f]{40}$/.test(normalized)) {
    throw new Error('Verifier address must be a valid 20-byte hex address.');
  }
  return `0x${normalized}`;
}

function toBytes(data: string, encoding: ZkProofEncoding): Uint8Array {
  if (encoding === 'utf8') {
    return new TextEncoder().encode(data);
  }

  if (encoding === 'base64') {
    return decodeBase64(data);
  }

  if (encoding === 'hex') {
    const normalized = normalizeHex(data);
    return decodeHex(normalized);
  }

  const maybeHex = normalizeHex(data);
  if (
    maybeHex.length > 0 &&
    maybeHex.length % 2 === 0 &&
    isHexString(maybeHex) &&
    data.trim().startsWith('0x')
  ) {
    return decodeHex(maybeHex);
  }

  return new TextEncoder().encode(data);
}

function bytesToHex(bytes: Uint8Array): string {
  return encodeHex(bytes);
}

async function toBytes32Word(value: string): Promise<string> {
  const normalized = normalizeHex(value);

  if (
    isHexString(normalized) &&
    normalized.length > 0 &&
    normalized.length <= 64
  ) {
    return encodeWord(normalized);
  }

  const digest = await crypto.subtle.digest(
    'SHA-256',
    new TextEncoder().encode(value)
  );
  return bytesToHex(new Uint8Array(digest));
}

function parseBooleanResult(returnData: string): boolean {
  const normalized = normalizeHex(returnData);
  if (normalized.length === 0) {
    return false;
  }

  if (!isHexString(normalized)) {
    throw new Error('eth_call returned non-hex result payload.');
  }

  const word = normalized.slice(-64);
  return BigInt(`0x${word}`) !== 0n;
}

/**
 * Builds the Verify Execution Calldata.
 */
export async function buildVerifyExecutionCalldata(
  input: ZkProofVerifierInput,
  config: EvmProofVerifierConfig
): Promise<string> {
  const methodSelector = normalizeMethodSelector(config.methodSelector);
  const proofEncoding = config.proofEncoding ?? 'auto';
  const proofBytes = toBytes(input.proof, proofEncoding);
  const proofHex = bytesToHex(proofBytes);
  const proofLengthHex = uintToHex(proofBytes.length);
  const proofPaddedHex = padHexRight(
    proofHex,
    Math.ceil(proofHex.length / 64) * 64
  );

  const publicSignalsHashWord = await toBytes32Word(input.publicSignalsHash);
  const programHashWord = await toBytes32Word(input.programHash);
  const vkHashWord = await toBytes32Word(input.vkHash);

  const dynamicOffsetWord = encodeWord(uintToHex(4 * 32));
  const proofLengthWord = encodeWord(proofLengthHex);

  return `0x${normalizeHex(
    methodSelector
  )}${dynamicOffsetWord}${publicSignalsHashWord}${programHashWord}${vkHashWord}${proofLengthWord}${proofPaddedHex}`;
}

interface JsonRpcSuccess {
  jsonrpc: string;
  id: number;
  result: string;
}

interface JsonRpcError {
  jsonrpc: string;
  id: number;
  error: {
    code: number;
    message: string;
    data?: unknown;
  };
}

function isJsonRpcSuccess(value: unknown): value is JsonRpcSuccess {
  if (typeof value !== 'object' || value === null) {
    return false;
  }
  const record = value as Record<string, unknown>;
  return typeof record.result === 'string';
}

function isJsonRpcError(value: unknown): value is JsonRpcError {
  if (typeof value !== 'object' || value === null) {
    return false;
  }
  const record = value as Record<string, unknown>;
  if (typeof record.error !== 'object' || record.error === null) {
    return false;
  }
  const errorRecord = record.error as Record<string, unknown>;
  return (
    typeof errorRecord.code === 'number' &&
    typeof errorRecord.message === 'string'
  );
}

/**
 * Handles the gnosis verify Proof Via Evm Rpc workflow.
 */
export async function verifyProofViaEvmRpc(
  input: ZkProofVerifierInput,
  config: EvmProofVerifierConfig
): Promise<boolean> {
  const rpcUrl = config.rpcUrl.trim();
  if (rpcUrl.length === 0) {
    throw new Error('rpcUrl is required for EVM proof verification.');
  }

  const to = normalizeAddress(config.verifierAddress);
  const data = await buildVerifyExecutionCalldata(input, config);
  const blockTag = config.blockTag ?? DEFAULT_BLOCK_TAG;
  const timeoutMs = config.timeoutMs ?? DEFAULT_TIMEOUT_MS;

  const requestBody = {
    jsonrpc: '2.0',
    id: 1,
    method: 'eth_call',
    params: [
      {
        to,
        data,
      },
      blockTag,
    ],
  };

  const controller = new AbortController();
  const timeoutHandle = setTimeout(() => controller.abort(), timeoutMs);

  let response: Response;
  try {
    response = await fetch(rpcUrl, {
      method: 'POST',
      headers: {
        'content-type': 'application/json',
      },
      body: JSON.stringify(requestBody),
      signal: controller.signal,
    });
  } finally {
    clearTimeout(timeoutHandle);
  }

  if (!response.ok) {
    throw new Error(`eth_call request failed with HTTP ${response.status}.`);
  }

  const parsed = (await response.json()) as unknown;
  if (isJsonRpcError(parsed)) {
    throw new Error(
      `eth_call JSON-RPC error ${parsed.error.code}: ${parsed.error.message}`
    );
  }

  if (!isJsonRpcSuccess(parsed)) {
    throw new Error('eth_call returned malformed JSON-RPC payload.');
  }

  return parseBooleanResult(parsed.result);
}

/**
 * Creates the Evm Proof Verifier.
 */
export function createEvmProofVerifier(
  config: EvmProofVerifierConfig
): ZkProofVerifier {
  return async (input: ZkProofVerifierInput): Promise<boolean> =>
    verifyProofViaEvmRpc(input, config);
}
