// @a0n/gnosis/distributed-inference-host/__tests__/style-transfer-imaging-wasm.test.ts
import { describe, expect, it } from 'vitest';
import {
  topologyAlphaStyleBackendFromImagingWasm,
  tryWasmMeshDepthConfidence,
} from '../src/style-transfer-imaging-wasm';

describe('style-transfer-imaging-wasm', () => {
  it('topologyAlphaStyleBackendFromImagingWasm maps wasm result to alpha + fragment', async () => {
    const backend = topologyAlphaStyleBackendFromImagingWasm({
      wasmTopologyMaskBuildRgbF32(rgb, w, h) {
        expect(rgb.length).toBe(w * h * 3);
        const alpha = new Float32Array(w * h);
        alpha.fill(0.42);
        return {
          alpha: () => alpha,
          provenance_json: () => JSON.stringify({ mock: true, w, h }),
        };
      },
    });
    const out = await backend.buildTopologyAlphaForStyle({
      data: new Float32Array(4 * 4 * 3).fill(0.5),
      width: 4,
      height: 4,
      channels: 3,
    });
    expect(out).not.toBeNull();
    expect(out!.alpha.every((v) => Math.abs(v - 0.42) < 1e-6)).toBe(true);
    expect(out!.provenanceFragment.mock).toBe(true);
  });

  it('topology backend returns null when wasm throws', async () => {
    const backend = topologyAlphaStyleBackendFromImagingWasm({
      wasmTopologyMaskBuildRgbF32() {
        throw new Error('boom');
      },
    });
    const out = await backend.buildTopologyAlphaForStyle({
      data: new Float32Array(12),
      width: 2,
      height: 2,
      channels: 3,
    });
    expect(out).toBeNull();
  });

  it('tryWasmMeshDepthConfidence returns null when export missing or throws', async () => {
    expect(
      await tryWasmMeshDepthConfidence(undefined, {
        data: new Float32Array(12),
        width: 2,
        height: 2,
        channels: 3,
        monocularCueBlend: 0,
      })
    ).toBeNull();
    expect(
      await tryWasmMeshDepthConfidence(
        {
          wasmTopologyMaskBuildRgbF32: () => ({
            alpha: () => new Float32Array(4),
            provenance_json: () => '{}',
          }),
        },
        {
          data: new Float32Array(12),
          width: 2,
          height: 2,
          channels: 3,
          monocularCueBlend: 0,
        }
      )
    ).toBeNull();
  });
});
