// @a0n/.ignored_gnosis/forkjoin-app/client/components/AeonLogo.tsx
import { memo, useCallback, useEffect, useRef } from '@a0n/raect';

/* ── hex geometry ── */
const R = 20;
const W = R * Math.sqrt(3);
const H = 2 * R;
const CX = 200;
const CY = 200;

function hexPath(cx: number, cy: number, r: number): string {
  const pts: string[] = [];
  for (let i = 0; i < 6; i++) {
    const angle = (Math.PI / 180) * (60 * i - 90);
    pts.push(`${cx + r * Math.cos(angle)},${cy + r * Math.sin(angle)}`);
  }
  return `M${pts.join('L')}Z`;
}

const HEXAGONS: {
  id: string;
  cx: number;
  cy: number;
  q: number;
  r: number;
  dist: number;
}[] = [];
for (let q = -2; q <= 2; q++) {
  const r1 = Math.max(-2, -q - 2);
  const r2 = Math.min(2, -q + 2);
  for (let row = r1; row <= r2; row++) {
    const x = CX + W * (q + row / 2);
    const y = CY + 1.5 * R * row;
    const dist = Math.max(Math.abs(q), Math.abs(row), Math.abs(q + row));
    HEXAGONS.push({ id: `hex_${q}_${row}`, cx: x, cy: y, q, r: row, dist });
  }
}

/* ── constants ── */
const FLICK_INTERVAL = 12_000;
const LERP_SPEED = 0.08;
const BREEZE_SENSITIVITY = 0.8;
const CRANK_RATIO = 0.15;

interface AeonLogoProps {
  className?: string;
  ariaLabel?: string;
  activitySignal?: number; // 0 to 1
}

function AeonLogoInner({
  className,
  ariaLabel = 'Aeon Foundation logo',
  activitySignal = 0,
}: AeonLogoProps) {
  const svgRef = useRef<SVGSVGElement>(null);
  const gRefs = useRef<(SVGGElement | null)[]>(new Array(19).fill(null));
  const crankRef = useRef<SVGGElement | null>(null);
  const targets = useRef(new Array(19).fill(0));
  const currents = useRef(new Array(19).fill(0));
  const crankTarget = useRef(0);
  const crankCurrent = useRef(0);
  const lastMouseAngle = useRef<number | null>(null);
  const lastMousePos = useRef({ x: 0, y: 0 });
  const rafId = useRef(0);
  const flickTimer = useRef<ReturnType<typeof setTimeout> | null>(null);

  /* ── activity: baseline rotation ── */
  useEffect(() => {
    if (activitySignal <= 0) return;

    const interval = setInterval(() => {
      for (let i = 0; i < 19; i++) {
        targets.current[i] += (Math.random() - 0.5) * 60 * activitySignal;
      }
    }, 2000 / (0.5 + activitySignal * 2));

    return () => clearInterval(interval);
  }, [activitySignal]);

  /* ── breeze & wave ── */
  const onMouseMove = useCallback((e: MouseEvent) => {
    const svg = svgRef.current;
    if (!svg) return;
    const rect = svg.getBoundingClientRect();
    const centerX = rect.left + rect.width / 2;
    const centerY = rect.top + rect.height / 2;

    const relX = e.clientX - centerX;
    const relY = e.clientY - centerY;

    const dx = e.clientX - lastMousePos.current.x;
    const dy = e.clientY - lastMousePos.current.y;
    const speed = Math.sqrt(dx * dx + dy * dy);

    // Map mouse position to SVG coordinates roughly
    const scale = 400 / rect.width;
    const svgMouseX = CX + relX * scale;
    const svgMouseY = CY + relY * scale;

    for (let i = 0; i < 19; i++) {
      const hex = HEXAGONS[i];
      const distToMouse = Math.sqrt(
        Math.pow(hex.cx - svgMouseX, 2) + Math.pow(hex.cy - svgMouseY, 2)
      );

      // Hexagons closer to mouse spin more
      const influence = Math.max(0, 1 - distToMouse / 150);
      if (influence > 0) {
        targets.current[i] +=
          speed *
          influence *
          BREEZE_SENSITIVITY *
          (Math.random() > 0.5 ? 1 : -1);
      }
    }

    /* ── crank: angular displacement of mouse around logo center ── */
    const angle = Math.atan2(relY, relX) * (180 / Math.PI);
    if (lastMouseAngle.current !== null) {
      let delta = angle - lastMouseAngle.current;
      if (delta > 180) delta -= 360;
      if (delta < -180) delta += 360;
      crankTarget.current += delta * CRANK_RATIO;
    }
    lastMouseAngle.current = angle;
    lastMousePos.current = { x: e.clientX, y: e.clientY };
  }, []);

  /* ── animation loop ── */
  useEffect(() => {
    let running = true;

    function tick() {
      if (!running) return;
      for (let i = 0; i < 19; i++) {
        const diff = targets.current[i] - currents.current[i];
        if (Math.abs(diff) > 0.05) {
          currents.current[i] += diff * LERP_SPEED;
        }
        const g = gRefs.current[i];
        if (g) {
          g.style.transform = `rotate(${currents.current[i].toFixed(
            2
          )}deg) scale(${1 + Math.sin(currents.current[i] * 0.05) * 0.1})`;
        }
      }

      const crankDiff = crankTarget.current - crankCurrent.current;
      if (Math.abs(crankDiff) > 0.01) {
        crankCurrent.current += crankDiff * LERP_SPEED;
      }
      if (crankRef.current) {
        crankRef.current.style.transform = `rotate(${crankCurrent.current.toFixed(
          2
        )}deg)`;
      }
      rafId.current = requestAnimationFrame(tick);
    }

    rafId.current = requestAnimationFrame(tick);
    window.addEventListener('mousemove', onMouseMove);

    return () => {
      running = false;
      cancelAnimationFrame(rafId.current);
      window.removeEventListener('mousemove', onMouseMove);
    };
  }, [onMouseMove]);

  /* ── ripple flick ── */
  const triggerRippleFlick = useCallback((startIndex: number) => {
    const startHex = HEXAGONS[startIndex];

    HEXAGONS.forEach((hex, i) => {
      // Calculate hexagonal grid distance
      const dq = hex.q - startHex.q;
      const dr = hex.r - startHex.r;
      const distance = Math.max(Math.abs(dq), Math.abs(dr), Math.abs(dq + dr));

      setTimeout(() => {
        const g = gRefs.current[i];
        if (!g) return;
        g.classList.remove('aeon-hex-flick');
        void g.getBoundingClientRect(); // force reflow
        g.classList.add('aeon-hex-flick');
        const onEnd = () => {
          g.classList.remove('aeon-hex-flick');
          g.removeEventListener('animationend', onEnd);
        };
        g.addEventListener('animationend', onEnd);
      }, distance * 150); // Delay based on distance for ripple effect
    });
  }, []);

  useEffect(() => {
    function scheduleFlick() {
      const jitter = (Math.random() - 0.5) * FLICK_INTERVAL * 0.3;
      flickTimer.current = setTimeout(() => {
        // Pick a random hexagon to start the ripple
        const idx = Math.floor(Math.random() * 19);
        triggerRippleFlick(idx);
        scheduleFlick();
      }, FLICK_INTERVAL + jitter);
    }

    flickTimer.current = setTimeout(() => {
      triggerRippleFlick(9); // 9 is usually the center (0,0) in our generation
      scheduleFlick();
    }, 2_000);

    return () => {
      if (flickTimer.current) clearTimeout(flickTimer.current);
    };
  }, [triggerRippleFlick]);

  return (
    <svg
      ref={svgRef}
      className={`aeon-logo${className ? ` ${className}` : ''}`}
      viewBox="0 0 400 400"
      role="img"
      aria-label={ariaLabel}
    >
      <style>{`
        .aeon-logo {
          display: inline-block;
          vertical-align: middle;
          color: currentColor;
          width: 100%;
          height: 100%;
        }
        @keyframes aeonHexFlick {
          0% { transform: rotate(0deg) scale(1); }
          15% { transform: rotate(-30deg) scale(0.9); }
          45% { transform: rotate(750deg) scale(1.1); }
          75% { transform: rotate(710deg) scale(1.05); }
          100% { transform: rotate(720deg) scale(1); }
        }
        .aeon-hex-flick {
          animation: aeonHexFlick 3.5s cubic-bezier(0.4, 0, 0.2, 1) forwards;
        }
        @media (prefers-reduced-motion: reduce) {
          .aeon-hex-flick { animation: none; }
          .aeon-logo g { transition: none !important; }
        }
      `}</style>
      <g
        ref={(el: SVGGElement | null) => {
          crankRef.current = el;
        }}
        style={{ transformOrigin: `${CX}px ${CY}px` }}
      >
        {HEXAGONS.map((hex, i) => (
          <g
            key={hex.id}
            ref={(el: SVGGElement | null) => {
              gRefs.current[i] = el;
            }}
            style={{
              transformOrigin: `${hex.cx}px ${hex.cy}px`,
            }}
          >
            <path
              d={hexPath(hex.cx, hex.cy, R)}
              fill="transparent"
              stroke="currentColor"
              strokeWidth="2"
              strokeLinejoin="round"
              onClick={() => triggerRippleFlick(i)}
              style={{ cursor: 'pointer', opacity: 0.8 + (2 - hex.dist) * 0.1 }}
            />
          </g>
        ))}
      </g>
    </svg>
  );
}

const AeonLogo = memo(AeonLogoInner);
AeonLogo.displayName = 'AeonLogo';
export default AeonLogo;
