// @a0n/gnosis/src/benchmarks/adversarial-controls-benchmark.ts
import { readFile } from 'node:fs/promises';

import { parseGgProgram } from '@a0n/aeon-logic';

import {
  DEFAULT_FOLD_TRAINING_TOPOLOGY_FILES,
  type FoldTrainingStrategy,
} from './fold-training-benchmark';
import { DEFAULT_MINI_MOE_ROUTING_TOPOLOGY_FILES } from './moe-routing-benchmark';
import {
  bootstrapMeanConfidenceInterval,
  createDeterministicRandom,
  mean,
  shuffleInPlace,
  stdev,
  type ConfidenceInterval,
} from './statistics';
import { lowerUfcsSource } from '../ufcs';

type AdversarialStrategy = FoldTrainingStrategy;
type AdversarialTaskId =
  | 'winner-affine-maxabs'
  | 'early-stop-routing-first-expert-short-budget'
  | 'early-stop-left-priority-short-budget';
type TaskSuccessCriterion = 'final-and-sample-efficiency' | 'sample-efficiency';

export interface AdversarialSeedMetrics {
  readonly seed: number;
  readonly finalTrainMeanSquaredError: number;
  readonly finalEvalMeanSquaredError: number;
  readonly exactWithinToleranceFraction: number;
  readonly learningCurveArea: number;
  readonly epochsToTolerance: number;
}

export interface AdversarialCheckpointMetric {
  readonly epoch: number;
  readonly meanEvalMeanSquaredError: number;
  readonly evalMeanSquaredErrorCi95: ConfidenceInterval;
}

export interface AdversarialStrategyReport {
  readonly seedMetrics: readonly AdversarialSeedMetrics[];
  readonly checkpoints: readonly AdversarialCheckpointMetric[];
  readonly meanFinalTrainMeanSquaredError: number;
  readonly stdevFinalTrainMeanSquaredError: number;
  readonly meanFinalEvalMeanSquaredError: number;
  readonly finalEvalMeanSquaredErrorCi95: ConfidenceInterval;
  readonly stdevFinalEvalMeanSquaredError: number;
  readonly meanExactWithinToleranceFraction: number;
  readonly exactWithinToleranceFractionCi95: ConfidenceInterval;
  readonly stdevExactWithinToleranceFraction: number;
  readonly meanLearningCurveArea: number;
  readonly learningCurveAreaCi95: ConfidenceInterval;
  readonly stdevLearningCurveArea: number;
  readonly meanEpochsToTolerance: number;
  readonly epochsToToleranceCi95: ConfidenceInterval;
  readonly stdevEpochsToTolerance: number;
}

export interface AdversarialTopologyMetrics {
  readonly nodeCount: number;
  readonly edgeCount: number;
  readonly structuralBeta1: number;
  readonly unitCount: number;
  readonly parameterCount: number;
}

export interface AdversarialTaskReport {
  readonly title: string;
  readonly description: string;
  readonly topologyFamily: 'fold-training' | 'moe-routing';
  readonly targetFamily:
    | 'winner-max-abs-affine'
    | 'x-priority-routing-short-budget'
    | 'left-priority-short-budget';
  readonly favoredStrategy: AdversarialStrategy;
  readonly successCriterion: TaskSuccessCriterion;
  readonly topology: AdversarialTopologyMetrics;
  readonly trainingBudget: {
    readonly seedCount: number;
    readonly epochs: number;
    readonly learningRate: number;
    readonly checkpointInterval: number;
    readonly exactPredictionTolerance: number;
  };
  readonly strategies: Record<AdversarialStrategy, AdversarialStrategyReport>;
  readonly rankingByFinalEvalMeanSquaredError: readonly AdversarialStrategy[];
  readonly rankingByLearningCurveArea: readonly AdversarialStrategy[];
  readonly favoredStrategyWinsFinal: boolean;
  readonly favoredStrategyWinsSampleEfficiency: boolean;
  readonly predictionRecovered: boolean;
}

export interface GnosisAdversarialControlsBenchmarkReport {
  readonly label: 'gnosis-adversarial-controls-benchmark-v1';
  readonly tasks: Record<AdversarialTaskId, AdversarialTaskReport>;
  readonly allAdversarialPredictionsRecovered: boolean;
}

interface AffineSample {
  readonly leftInput: number;
  readonly rightInput: number;
  readonly target: number;
}

interface RoutedSample {
  readonly leftInput: number;
  readonly rightInput: number;
  readonly target: number;
}

interface AffineParameters {
  leftWeight: number;
  leftBias: number;
  rightWeight: number;
  rightBias: number;
}

interface ExpertParameters {
  gateWeight: number;
  gateBias: number;
  valueWeight: number;
  valueBias: number;
}

interface AffinePrediction {
  readonly leftValue: number;
  readonly rightValue: number;
  readonly prediction: number;
  readonly selectedBranch: 'left' | 'right';
}

interface ExpertSpec {
  readonly signedProjection: (sample: RoutedSample) => number;
  readonly axisProjection: (sample: RoutedSample) => number;
}

interface ExpertActivation {
  readonly gate: number;
  readonly value: number;
  readonly output: number;
  readonly signedInput: number;
  readonly axisInput: number;
}

interface RoutedPrediction {
  readonly prediction: number;
  readonly selectedExpertIndex: number;
  readonly expertActivations: readonly ExpertActivation[];
}

interface AdversarialTaskConfig {
  readonly seedCount: number;
  readonly epochs: number;
  readonly learningRate: number;
  readonly checkpointInterval: number;
  readonly exactPredictionTolerance: number;
}

const adversarialStrategies = Object.keys(
  DEFAULT_FOLD_TRAINING_TOPOLOGY_FILES
) as AdversarialStrategy[];

const expertSpecs: readonly ExpertSpec[] = [
  {
    signedProjection: (sample) => sample.leftInput,
    axisProjection: (sample) => sample.leftInput,
  },
  {
    signedProjection: (sample) => -sample.leftInput,
    axisProjection: (sample) => sample.leftInput,
  },
  {
    signedProjection: (sample) => sample.rightInput,
    axisProjection: (sample) => sample.rightInput,
  },
  {
    signedProjection: (sample) => -sample.rightInput,
    axisProjection: (sample) => sample.rightInput,
  },
] as const;

function sigmoid(value: number): number {
  return 1 / (1 + Math.exp(-value));
}

function countConnectedComponents(
  nodeIds: readonly string[],
  edges: readonly {
    readonly sourceIds: readonly string[];
    readonly targetIds: readonly string[];
  }[]
): number {
  const neighbors = new Map<string, Set<string>>();
  for (const nodeId of nodeIds) {
    neighbors.set(nodeId, new Set<string>());
  }

  for (const edge of edges) {
    for (const sourceId of edge.sourceIds) {
      const sourceNeighbors = neighbors.get(sourceId);
      if (!sourceNeighbors) {
        continue;
      }
      for (const targetId of edge.targetIds) {
        sourceNeighbors.add(targetId);
        neighbors.get(targetId)?.add(sourceId);
      }
    }
  }

  let components = 0;
  const visited = new Set<string>();
  for (const nodeId of nodeIds) {
    if (visited.has(nodeId)) {
      continue;
    }
    components++;
    const frontier = [nodeId];
    while (frontier.length > 0) {
      const current = frontier.pop();
      if (current === undefined || visited.has(current)) {
        continue;
      }
      visited.add(current);
      for (const neighbor of neighbors.get(current) ?? []) {
        if (!visited.has(neighbor)) {
          frontier.push(neighbor);
        }
      }
    }
  }

  return Math.max(1, components);
}

function structuralBeta1(
  nodeIds: readonly string[],
  edges: readonly {
    readonly sourceIds: readonly string[];
    readonly targetIds: readonly string[];
  }[]
): number {
  const expandedEdgeCount = edges.reduce(
    (sum, edge) => sum + edge.sourceIds.length * edge.targetIds.length,
    0
  );
  return Math.max(
    0,
    expandedEdgeCount -
      nodeIds.length +
      countConnectedComponents(nodeIds, edges)
  );
}

async function loadAffineTopologyMetrics(): Promise<AdversarialTopologyMetrics> {
  const source = await readFile(
    DEFAULT_FOLD_TRAINING_TOPOLOGY_FILES.linear,
    'utf8'
  );
  const program = parseGgProgram(lowerUfcsSource(source));
  const branchNodes = program.nodes.filter((node) =>
    node.labels.includes('AffineBranch')
  );
  return {
    nodeCount: program.nodes.length,
    edgeCount: program.edges.length,
    structuralBeta1: structuralBeta1(
      program.nodes.map((node) => node.id),
      program.edges.map((edge) => ({
        sourceIds: edge.sourceIds,
        targetIds: edge.targetIds,
      }))
    ),
    unitCount: branchNodes.length,
    parameterCount: branchNodes.length * 2,
  };
}

async function loadRoutedTopologyMetrics(): Promise<AdversarialTopologyMetrics> {
  const source = await readFile(
    DEFAULT_MINI_MOE_ROUTING_TOPOLOGY_FILES.linear,
    'utf8'
  );
  const program = parseGgProgram(lowerUfcsSource(source));
  const expertNodes = program.nodes.filter((node) =>
    node.labels.includes('MoEExpert')
  );
  return {
    nodeCount: program.nodes.length,
    edgeCount: program.edges.length,
    structuralBeta1: structuralBeta1(
      program.nodes.map((node) => node.id),
      program.edges.map((edge) => ({
        sourceIds: edge.sourceIds,
        targetIds: edge.targetIds,
      }))
    ),
    unitCount: expertNodes.length,
    parameterCount: expertNodes.reduce((sum, node) => {
      const parameterCount = Number(node.properties.parameters ?? 0);
      return sum + (Number.isFinite(parameterCount) ? parameterCount : 0);
    }, 0),
  };
}

function initializeAffineParameters(seed: number): AffineParameters {
  const random = createDeterministicRandom(seed);
  const span = 0.8;
  const offset = -0.4;
  return {
    leftWeight: offset + random() * span,
    leftBias: offset + random() * span,
    rightWeight: offset + random() * span,
    rightBias: offset + random() * span,
  };
}

function initializeRoutedParameters(seed: number): ExpertParameters[] {
  const random = createDeterministicRandom(seed);
  return expertSpecs.map(() => ({
    gateWeight: (random() - 0.5) * 1.4,
    gateBias: (random() - 0.5) * 0.5,
    valueWeight: (random() - 0.5) * 1.0,
    valueBias: (random() - 0.5) * 0.4,
  }));
}

function buildAffineSamples(
  leftAxis: readonly number[],
  rightAxis: readonly number[],
  target: (leftInput: number, rightInput: number) => number
): AffineSample[] {
  return leftAxis.flatMap((leftInput) =>
    rightAxis.map((rightInput) => ({
      leftInput,
      rightInput,
      target: target(leftInput, rightInput),
    }))
  );
}

function buildRoutedSamples(
  leftAxis: readonly number[],
  rightAxis: readonly number[],
  target: (leftInput: number, rightInput: number) => number
): RoutedSample[] {
  return leftAxis.flatMap((leftInput) =>
    rightAxis.map((rightInput) => ({
      leftInput,
      rightInput,
      target: target(leftInput, rightInput),
    }))
  );
}

function predictAffine(
  strategy: AdversarialStrategy,
  parameters: AffineParameters,
  sample: AffineSample
): AffinePrediction {
  const leftValue =
    parameters.leftWeight * sample.leftInput + parameters.leftBias;
  const rightValue =
    parameters.rightWeight * sample.rightInput + parameters.rightBias;

  if (strategy === 'linear') {
    return {
      leftValue,
      rightValue,
      prediction: leftValue + rightValue,
      selectedBranch: 'left',
    };
  }

  if (strategy === 'early-stop') {
    return {
      leftValue,
      rightValue,
      prediction: leftValue,
      selectedBranch: 'left',
    };
  }

  const selectedBranch =
    Math.abs(leftValue) >= Math.abs(rightValue) ? 'left' : 'right';
  return {
    leftValue,
    rightValue,
    prediction: selectedBranch === 'left' ? leftValue : rightValue,
    selectedBranch,
  };
}

function updateAffineParameters(
  strategy: AdversarialStrategy,
  parameters: AffineParameters,
  sample: AffineSample,
  learningRate: number
): void {
  const prediction = predictAffine(strategy, parameters, sample);
  const error = prediction.prediction - sample.target;

  if (strategy === 'linear') {
    parameters.leftWeight -= learningRate * error * sample.leftInput;
    parameters.leftBias -= learningRate * error;
    parameters.rightWeight -= learningRate * error * sample.rightInput;
    parameters.rightBias -= learningRate * error;
    return;
  }

  if (prediction.selectedBranch === 'left') {
    parameters.leftWeight -= learningRate * error * sample.leftInput;
    parameters.leftBias -= learningRate * error;
    return;
  }

  parameters.rightWeight -= learningRate * error * sample.rightInput;
  parameters.rightBias -= learningRate * error;
}

function evaluateAffineSamples(
  strategy: AdversarialStrategy,
  parameters: AffineParameters,
  samples: readonly AffineSample[],
  tolerance: number
): {
  readonly meanSquaredError: number;
  readonly exactWithinToleranceFraction: number;
} {
  if (samples.length === 0) {
    return {
      meanSquaredError: 0,
      exactWithinToleranceFraction: 0,
    };
  }

  let squaredErrorSum = 0;
  let exactCount = 0;
  for (const sample of samples) {
    const error =
      predictAffine(strategy, parameters, sample).prediction - sample.target;
    squaredErrorSum += error * error;
    if (Math.abs(error) <= tolerance) {
      exactCount++;
    }
  }

  return {
    meanSquaredError: squaredErrorSum / samples.length,
    exactWithinToleranceFraction: exactCount / samples.length,
  };
}

function predictRouted(
  strategy: AdversarialStrategy,
  parameters: readonly ExpertParameters[],
  sample: RoutedSample
): RoutedPrediction {
  const expertActivations = expertSpecs.map((expert, index) => {
    const parameter = parameters[index];
    if (!parameter) {
      throw new Error(`Missing routed parameters for expert ${index}`);
    }

    const signedInput = expert.signedProjection(sample);
    const axisInput = expert.axisProjection(sample);
    const gate = sigmoid(
      parameter.gateWeight * signedInput + parameter.gateBias
    );
    const value = parameter.valueWeight * axisInput + parameter.valueBias;
    return {
      gate,
      value,
      output: gate * value,
      signedInput,
      axisInput,
    };
  });

  let selectedExpertIndex = 0;
  for (let index = 1; index < expertActivations.length; index++) {
    const selected = expertActivations[selectedExpertIndex];
    const candidate = expertActivations[index];
    if (
      candidate &&
      selected &&
      Math.abs(candidate.output) > Math.abs(selected.output)
    ) {
      selectedExpertIndex = index;
    }
  }

  if (strategy === 'linear') {
    return {
      prediction: expertActivations.reduce(
        (sum, activation) => sum + activation.output,
        0
      ),
      selectedExpertIndex,
      expertActivations,
    };
  }

  if (strategy === 'early-stop') {
    return {
      prediction: expertActivations[0]?.output ?? 0,
      selectedExpertIndex: 0,
      expertActivations,
    };
  }

  return {
    prediction: expertActivations[selectedExpertIndex]?.output ?? 0,
    selectedExpertIndex,
    expertActivations,
  };
}

function updateRoutedParameters(
  strategy: AdversarialStrategy,
  parameters: ExpertParameters[],
  sample: RoutedSample,
  learningRate: number
): void {
  const prediction = predictRouted(strategy, parameters, sample);
  const error = prediction.prediction - sample.target;

  for (let index = 0; index < parameters.length; index++) {
    if (
      strategy === 'winner-take-all' &&
      index !== prediction.selectedExpertIndex
    ) {
      continue;
    }
    if (strategy === 'early-stop' && index !== 0) {
      continue;
    }

    const parameter = parameters[index];
    const activation = prediction.expertActivations[index];
    if (!parameter || !activation) {
      continue;
    }

    const gatePrime = activation.gate * (1 - activation.gate);
    parameter.valueWeight -=
      learningRate * error * activation.gate * activation.axisInput;
    parameter.valueBias -= learningRate * error * activation.gate;
    parameter.gateWeight -=
      learningRate *
      error *
      gatePrime *
      activation.signedInput *
      activation.value;
    parameter.gateBias -= learningRate * error * gatePrime * activation.value;
  }
}

function evaluateRoutedSamples(
  strategy: AdversarialStrategy,
  parameters: readonly ExpertParameters[],
  samples: readonly RoutedSample[],
  tolerance: number
): {
  readonly meanSquaredError: number;
  readonly exactWithinToleranceFraction: number;
} {
  if (samples.length === 0) {
    return {
      meanSquaredError: 0,
      exactWithinToleranceFraction: 0,
    };
  }

  let squaredErrorSum = 0;
  let exactCount = 0;
  for (const sample of samples) {
    const error =
      predictRouted(strategy, parameters, sample).prediction - sample.target;
    squaredErrorSum += error * error;
    if (Math.abs(error) <= tolerance) {
      exactCount++;
    }
  }

  return {
    meanSquaredError: squaredErrorSum / samples.length,
    exactWithinToleranceFraction: exactCount / samples.length,
  };
}

function summarizeCheckpoints(
  perSeedCheckpoints: readonly (readonly {
    readonly epoch: number;
    readonly evalMeanSquaredError: number;
  }[])[],
  ciSeedBase: number
): AdversarialCheckpointMetric[] {
  const epochCount = perSeedCheckpoints[0]?.length ?? 0;
  return Array.from({ length: epochCount }, (_, checkpointIndex) => {
    const epoch = perSeedCheckpoints[0]?.[checkpointIndex]?.epoch ?? 0;
    const values = perSeedCheckpoints.map(
      (seedCheckpoints) =>
        seedCheckpoints[checkpointIndex]?.evalMeanSquaredError ?? 0
    );
    return {
      epoch,
      meanEvalMeanSquaredError: mean(values),
      evalMeanSquaredErrorCi95: bootstrapMeanConfidenceInterval(values, {
        seed: ciSeedBase + checkpointIndex,
      }),
    };
  });
}

function summarizeStrategyReport(
  metrics: readonly AdversarialSeedMetrics[],
  checkpoints: readonly (readonly {
    readonly epoch: number;
    readonly evalMeanSquaredError: number;
  }[])[],
  ciSeedBase: number
): AdversarialStrategyReport {
  const finalEvalValues = metrics.map(
    (entry) => entry.finalEvalMeanSquaredError
  );
  const exactValues = metrics.map(
    (entry) => entry.exactWithinToleranceFraction
  );
  const learningCurveAreas = metrics.map((entry) => entry.learningCurveArea);
  const epochsToTolerance = metrics.map((entry) => entry.epochsToTolerance);

  return {
    seedMetrics: metrics,
    checkpoints: summarizeCheckpoints(checkpoints, ciSeedBase),
    meanFinalTrainMeanSquaredError: mean(
      metrics.map((entry) => entry.finalTrainMeanSquaredError)
    ),
    stdevFinalTrainMeanSquaredError: stdev(
      metrics.map((entry) => entry.finalTrainMeanSquaredError)
    ),
    meanFinalEvalMeanSquaredError: mean(finalEvalValues),
    finalEvalMeanSquaredErrorCi95: bootstrapMeanConfidenceInterval(
      finalEvalValues,
      {
        seed: ciSeedBase,
      }
    ),
    stdevFinalEvalMeanSquaredError: stdev(finalEvalValues),
    meanExactWithinToleranceFraction: mean(exactValues),
    exactWithinToleranceFractionCi95: bootstrapMeanConfidenceInterval(
      exactValues,
      {
        seed: ciSeedBase ^ 0x00123456,
      }
    ),
    stdevExactWithinToleranceFraction: stdev(exactValues),
    meanLearningCurveArea: mean(learningCurveAreas),
    learningCurveAreaCi95: bootstrapMeanConfidenceInterval(learningCurveAreas, {
      seed: ciSeedBase ^ 0x00234567,
    }),
    stdevLearningCurveArea: stdev(learningCurveAreas),
    meanEpochsToTolerance: mean(epochsToTolerance),
    epochsToToleranceCi95: bootstrapMeanConfidenceInterval(epochsToTolerance, {
      seed: ciSeedBase ^ 0x00345678,
    }),
    stdevEpochsToTolerance: stdev(epochsToTolerance),
  };
}

function rankStrategies(
  strategies: Record<AdversarialStrategy, AdversarialStrategyReport>,
  key: 'meanFinalEvalMeanSquaredError' | 'meanLearningCurveArea'
): AdversarialStrategy[] {
  return [...adversarialStrategies].sort(
    (left, right) => strategies[left][key] - strategies[right][key]
  );
}

function collectAffineStrategyReport(
  strategy: AdversarialStrategy,
  config: AdversarialTaskConfig,
  trainSamples: readonly AffineSample[],
  evalSamples: readonly AffineSample[]
): AdversarialStrategyReport {
  const metrics: AdversarialSeedMetrics[] = [];
  const checkpoints: Array<
    readonly {
      readonly epoch: number;
      readonly evalMeanSquaredError: number;
    }[]
  > = [];

  for (let seed = 1; seed <= config.seedCount; seed++) {
    const random = createDeterministicRandom(0xb300 + seed * 17);
    const parameters = initializeAffineParameters(seed);
    const shuffledSamples = [...trainSamples];
    const seedCheckpoints: { epoch: number; evalMeanSquaredError: number }[] =
      [];
    let epochsToTolerance = config.epochs + 1;

    for (let epoch = 1; epoch <= config.epochs; epoch++) {
      shuffleInPlace(shuffledSamples, random);
      for (const sample of shuffledSamples) {
        updateAffineParameters(
          strategy,
          parameters,
          sample,
          config.learningRate
        );
      }

      if (epoch % config.checkpointInterval === 0 || epoch === config.epochs) {
        const checkpointMetrics = evaluateAffineSamples(
          strategy,
          parameters,
          evalSamples,
          config.exactPredictionTolerance
        );
        seedCheckpoints.push({
          epoch,
          evalMeanSquaredError: checkpointMetrics.meanSquaredError,
        });
        if (
          epochsToTolerance === config.epochs + 1 &&
          checkpointMetrics.meanSquaredError <= config.exactPredictionTolerance
        ) {
          epochsToTolerance = epoch;
        }
      }
    }

    const trainMetrics = evaluateAffineSamples(
      strategy,
      parameters,
      trainSamples,
      config.exactPredictionTolerance
    );
    const evalMetrics = evaluateAffineSamples(
      strategy,
      parameters,
      evalSamples,
      config.exactPredictionTolerance
    );

    metrics.push({
      seed,
      finalTrainMeanSquaredError: trainMetrics.meanSquaredError,
      finalEvalMeanSquaredError: evalMetrics.meanSquaredError,
      exactWithinToleranceFraction: evalMetrics.exactWithinToleranceFraction,
      learningCurveArea: mean(
        seedCheckpoints.map((entry) => entry.evalMeanSquaredError)
      ),
      epochsToTolerance,
    });
    checkpoints.push(seedCheckpoints);
  }

  return summarizeStrategyReport(
    metrics,
    checkpoints,
    0x610000 + strategy.length * 137
  );
}

function collectRoutedStrategyReport(
  strategy: AdversarialStrategy,
  config: AdversarialTaskConfig,
  trainSamples: readonly RoutedSample[],
  evalSamples: readonly RoutedSample[]
): AdversarialStrategyReport {
  const metrics: AdversarialSeedMetrics[] = [];
  const checkpoints: Array<
    readonly {
      readonly epoch: number;
      readonly evalMeanSquaredError: number;
    }[]
  > = [];

  for (let seed = 1; seed <= config.seedCount; seed++) {
    const random = createDeterministicRandom(0xc400 + seed * 19);
    const parameters = initializeRoutedParameters(seed);
    const shuffledSamples = [...trainSamples];
    const seedCheckpoints: { epoch: number; evalMeanSquaredError: number }[] =
      [];
    let epochsToTolerance = config.epochs + 1;

    for (let epoch = 1; epoch <= config.epochs; epoch++) {
      shuffleInPlace(shuffledSamples, random);
      for (const sample of shuffledSamples) {
        updateRoutedParameters(
          strategy,
          parameters,
          sample,
          config.learningRate
        );
      }

      if (epoch % config.checkpointInterval === 0 || epoch === config.epochs) {
        const checkpointMetrics = evaluateRoutedSamples(
          strategy,
          parameters,
          evalSamples,
          config.exactPredictionTolerance
        );
        seedCheckpoints.push({
          epoch,
          evalMeanSquaredError: checkpointMetrics.meanSquaredError,
        });
        if (
          epochsToTolerance === config.epochs + 1 &&
          checkpointMetrics.meanSquaredError <= config.exactPredictionTolerance
        ) {
          epochsToTolerance = epoch;
        }
      }
    }

    const trainMetrics = evaluateRoutedSamples(
      strategy,
      parameters,
      trainSamples,
      config.exactPredictionTolerance
    );
    const evalMetrics = evaluateRoutedSamples(
      strategy,
      parameters,
      evalSamples,
      config.exactPredictionTolerance
    );

    metrics.push({
      seed,
      finalTrainMeanSquaredError: trainMetrics.meanSquaredError,
      finalEvalMeanSquaredError: evalMetrics.meanSquaredError,
      exactWithinToleranceFraction: evalMetrics.exactWithinToleranceFraction,
      learningCurveArea: mean(
        seedCheckpoints.map((entry) => entry.evalMeanSquaredError)
      ),
      epochsToTolerance,
    });
    checkpoints.push(seedCheckpoints);
  }

  return summarizeStrategyReport(
    metrics,
    checkpoints,
    0x710000 + strategy.length * 137
  );
}

function buildTaskReport(
  title: string,
  description: string,
  topologyFamily: 'fold-training' | 'moe-routing',
  targetFamily: AdversarialTaskReport['targetFamily'],
  favoredStrategy: AdversarialStrategy,
  successCriterion: TaskSuccessCriterion,
  topology: AdversarialTopologyMetrics,
  trainingBudget: AdversarialTaskConfig,
  strategies: Record<AdversarialStrategy, AdversarialStrategyReport>
): AdversarialTaskReport {
  const rankingByFinalEvalMeanSquaredError = rankStrategies(
    strategies,
    'meanFinalEvalMeanSquaredError'
  );
  const rankingByLearningCurveArea = rankStrategies(
    strategies,
    'meanLearningCurveArea'
  );
  const favoredStrategyWinsFinal =
    rankingByFinalEvalMeanSquaredError[0] === favoredStrategy;
  const favoredStrategyWinsSampleEfficiency =
    rankingByLearningCurveArea[0] === favoredStrategy;
  const predictionRecovered =
    successCriterion === 'final-and-sample-efficiency'
      ? favoredStrategyWinsFinal && favoredStrategyWinsSampleEfficiency
      : favoredStrategyWinsSampleEfficiency;

  return {
    title,
    description,
    topologyFamily,
    targetFamily,
    favoredStrategy,
    successCriterion,
    topology,
    trainingBudget,
    strategies,
    rankingByFinalEvalMeanSquaredError,
    rankingByLearningCurveArea,
    favoredStrategyWinsFinal,
    favoredStrategyWinsSampleEfficiency,
    predictionRecovered,
  };
}

function makeDefaultAdversarialConfigs(): {
  readonly winnerAffine: AdversarialTaskConfig;
  readonly winnerRouted: AdversarialTaskConfig;
  readonly earlyStopShortBudget: AdversarialTaskConfig;
} {
  return {
    winnerAffine: {
      seedCount: 6,
      epochs: 480,
      learningRate: 0.04,
      checkpointInterval: 24,
      exactPredictionTolerance: 0.08,
    },
    winnerRouted: {
      seedCount: 8,
      epochs: 24,
      learningRate: 0.03,
      checkpointInterval: 8,
      exactPredictionTolerance: 0.08,
    },
    earlyStopShortBudget: {
      seedCount: 6,
      epochs: 24,
      learningRate: 0.03,
      checkpointInterval: 4,
      exactPredictionTolerance: 0.08,
    },
  };
}

/**
 * Handles the gnosis run Gnosis Adversarial Controls Benchmark workflow.
 */
export async function runGnosisAdversarialControlsBenchmark(): Promise<GnosisAdversarialControlsBenchmarkReport> {
  const configs = makeDefaultAdversarialConfigs();
  const [affineTopology, routedTopology] = await Promise.all([
    loadAffineTopologyMetrics(),
    loadRoutedTopologyMetrics(),
  ]);

  const winnerAffineTrain = buildAffineSamples(
    [-1.5, -1.0, -0.5, 0.5, 1.0, 1.5],
    [-1.5, -1.0, -0.5, 0.5, 1.0, 1.5],
    (leftInput, rightInput) =>
      Math.abs(leftInput) >= Math.abs(rightInput) ? leftInput : rightInput
  );
  const winnerAffineEval = buildAffineSamples(
    [-1.25, -0.75, -0.25, 0.25, 0.75, 1.25],
    [-1.25, -0.75, -0.25, 0.25, 0.75, 1.25],
    (leftInput, rightInput) =>
      Math.abs(leftInput) >= Math.abs(rightInput) ? leftInput : rightInput
  );
  const winnerAffineStrategies = Object.fromEntries(
    adversarialStrategies.map((strategy) => [
      strategy,
      collectAffineStrategyReport(
        strategy,
        configs.winnerAffine,
        winnerAffineTrain,
        winnerAffineEval
      ),
    ])
  ) as Record<AdversarialStrategy, AdversarialStrategyReport>;

  const winnerRoutedTrain = [0.2, 0.4, 0.6, 0.8, 1.0, 1.2, 1.4].flatMap(
    (leftInput) => [
      { leftInput, rightInput: leftInput, target: leftInput },
      { leftInput, rightInput: leftInput * 2, target: leftInput },
    ]
  );
  const winnerRoutedEval = buildRoutedSamples(
    [0.25, 0.45, 0.65, 0.85, 1.05, 1.25],
    [-1.0, -0.5, 0.0, 0.5, 1.0],
    (leftInput) => leftInput
  );
  const winnerRoutedStrategies = Object.fromEntries(
    adversarialStrategies.map((strategy) => [
      strategy,
      collectRoutedStrategyReport(
        strategy,
        configs.winnerRouted,
        winnerRoutedTrain,
        winnerRoutedEval
      ),
    ])
  ) as Record<AdversarialStrategy, AdversarialStrategyReport>;

  const earlyStopTrain = [-1.5, -1.0, -0.5, 0.5, 1.0, 1.5].flatMap(
    (leftInput) => [
      { leftInput, rightInput: leftInput, target: leftInput },
      { leftInput, rightInput: leftInput * 2, target: leftInput },
    ]
  );
  const earlyStopEval = buildAffineSamples(
    [-1.25, -0.75, -0.25, 0.25, 0.75, 1.25],
    [-2.0, -1.0, 0.0, 1.0, 2.0],
    (leftInput) => leftInput
  );
  const earlyStopStrategies = Object.fromEntries(
    adversarialStrategies.map((strategy) => [
      strategy,
      collectAffineStrategyReport(
        strategy,
        configs.earlyStopShortBudget,
        earlyStopTrain,
        earlyStopEval
      ),
    ])
  ) as Record<AdversarialStrategy, AdversarialStrategyReport>;

  const tasks: Record<AdversarialTaskId, AdversarialTaskReport> = {
    'winner-affine-maxabs': buildTaskReport(
      'Winner-aligned affine max-abs task',
      'Keeps the affine two-branch topology fixed but switches the target to signed max-by-magnitude, where winner-selection is the correct recombination rule.',
      'fold-training',
      'winner-max-abs-affine',
      'winner-take-all',
      'final-and-sample-efficiency',
      affineTopology,
      configs.winnerAffine,
      winnerAffineStrategies
    ),
    'early-stop-routing-first-expert-short-budget': buildTaskReport(
      'Early-stop aligned routed first-expert task',
      'Keeps the four-expert routed topology fixed, correlates the y-axis with x during training, and evaluates on independent y so always taking the first x-positive expert should pay off under a short budget.',
      'moe-routing',
      'x-priority-routing-short-budget',
      'early-stop',
      'sample-efficiency',
      routedTopology,
      configs.winnerRouted,
      winnerRoutedStrategies
    ),
    'early-stop-left-priority-short-budget': buildTaskReport(
      'Early-stop aligned left-priority task',
      'Keeps the affine topology fixed, injects large distractor right-branch magnitudes, and measures a short training budget where always taking the first branch should pay off in sample efficiency.',
      'fold-training',
      'left-priority-short-budget',
      'early-stop',
      'sample-efficiency',
      affineTopology,
      configs.earlyStopShortBudget,
      earlyStopStrategies
    ),
  };

  return {
    label: 'gnosis-adversarial-controls-benchmark-v1',
    tasks,
    allAdversarialPredictionsRecovered: Object.values(tasks).every(
      (task) => task.predictionRecovered
    ),
  };
}

function renderTaskMarkdown(
  taskId: string,
  task: AdversarialTaskReport
): string {
  const lines: string[] = [];
  lines.push(`## ${task.title}`);
  lines.push('');
  lines.push(`- Task id: \`${taskId}\``);
  lines.push(`- Description: ${task.description}`);
  lines.push(`- Topology family: \`${task.topologyFamily}\``);
  lines.push(`- Target family: \`${task.targetFamily}\``);
  lines.push(`- Favored strategy: \`${task.favoredStrategy}\``);
  lines.push(`- Success criterion: \`${task.successCriterion}\``);
  lines.push(
    `- Prediction recovered: \`${task.predictionRecovered ? 'yes' : 'no'}\``
  );
  lines.push('');
  lines.push(
    '| Strategy | Final eval MSE | Final eval 95% CI | Exact | Learning-curve area | Area 95% CI | Mean epochs-to-tolerance | Final rank | AUC rank |'
  );
  lines.push('| --- | ---: | ---: | ---: | ---: | ---: | ---: | ---: | ---: |');
  for (const strategy of adversarialStrategies) {
    const report = task.strategies[strategy];
    lines.push(
      `| \`${strategy}\` | ${report.meanFinalEvalMeanSquaredError.toFixed(
        3
      )} | [${report.finalEvalMeanSquaredErrorCi95.low.toFixed(
        3
      )}, ${report.finalEvalMeanSquaredErrorCi95.high.toFixed(
        3
      )}] | ${report.meanExactWithinToleranceFraction.toFixed(
        3
      )} | ${report.meanLearningCurveArea.toFixed(
        3
      )} | [${report.learningCurveAreaCi95.low.toFixed(
        3
      )}, ${report.learningCurveAreaCi95.high.toFixed(
        3
      )}] | ${report.meanEpochsToTolerance.toFixed(1)} | ${
        task.rankingByFinalEvalMeanSquaredError.indexOf(strategy) + 1
      } | ${task.rankingByLearningCurveArea.indexOf(strategy) + 1} |`
    );
  }
  lines.push('');
  return lines.join('\n');
}

/**
 * Handles the gnosis render Gnosis Adversarial Controls Benchmark Markdown workflow.
 */
export function renderGnosisAdversarialControlsBenchmarkMarkdown(
  report: GnosisAdversarialControlsBenchmarkReport
): string {
  const lines: string[] = [];
  lines.push('# Gnosis Adversarial Controls Benchmark');
  lines.push('');
  lines.push(`- Label: \`${report.label}\``);
  lines.push(
    `- All adversarial predictions recovered: \`${
      report.allAdversarialPredictionsRecovered ? 'yes' : 'no'
    }\``
  );
  lines.push('');
  for (const [taskId, task] of Object.entries(report.tasks)) {
    lines.push(renderTaskMarkdown(taskId, task));
    lines.push('');
  }
  lines.push(
    'Interpretation: these controls flip the story around. They keep the parameter-matched Chapter 17 topologies fixed, but choose targets where sparse nonlinear selection is the right inductive bias or where early stopping should help under a tight budget. The point is to show a boundary, not a one-sided preference for linearity.'
  );
  lines.push('');
  return `${lines.join('\n')}\n`;
}
