// @a0n/gnosis/src/benchmarks/aeon-framed-transformer-benchmark.ts
import { readFile } from 'node:fs/promises';
import { fileURLToPath } from 'node:url';

import {
  FIN,
  FOLD,
  FORK,
  FlowCodec,
  FrameReassembler,
  RACE,
  type FlowFrame,
} from '@a0n/aeon/flow';
import { parseGgProgram } from '@a0n/aeon-logic';
import { lowerUfcsSource } from '../ufcs';
import {
  bootstrapMeanConfidenceInterval,
  mean,
  stdev,
  createDeterministicRandom,
  type ConfidenceInterval,
} from './statistics';

export type AeonFramedTransformerStrategy =
  | 'linear'
  | 'winner-take-all'
  | 'early-stop';

export interface AeonFramedTransformerConfig {
  readonly label: 'gnosis-aeon-framed-transformer-benchmark-v1';
  readonly seedCount: number;
  readonly epochs: number;
  readonly learningRate: number;
  readonly exactPredictionTolerance: number;
  readonly dualActiveThreshold: number;
  readonly frameStageCount: number;
  readonly trainAxis: readonly number[];
  readonly evalAxis: readonly number[];
  readonly samplePoints: readonly (readonly [number, number])[];
}

export interface AeonFramedTransformerSeedMetrics {
  readonly seed: number;
  readonly trainMeanSquaredError: number;
  readonly evalMeanSquaredError: number;
  readonly exactWithinToleranceFraction: number;
  readonly dualActiveRegionMeanAbsoluteError: number;
  readonly codecRoundTripExactFraction: number;
  readonly reassemblyExactFraction: number;
  readonly foldInvarianceExactFraction: number;
  readonly frameCount: number;
  readonly payloadBytes: number;
}

export interface AeonFramedTransformerSamplePrediction {
  readonly leftInput: number;
  readonly rightInput: number;
  readonly target: number;
  readonly meanPrediction: number;
  readonly predictionStddev: number;
  readonly meanFrameCount: number;
  readonly meanPayloadBytes: number;
}

export interface AeonFramedTransformerStrategyReport {
  readonly topologyPath: string;
  readonly topologyNodeCount: number;
  readonly topologyEdgeCount: number;
  readonly structuralBeta1: number;
  readonly transformerletCount: number;
  readonly rotationStageCount: number;
  readonly parameterCount: number;
  readonly streamCountPerSample: number;
  readonly seedMetrics: readonly AeonFramedTransformerSeedMetrics[];
  readonly meanTrainMeanSquaredError: number;
  readonly stdevTrainMeanSquaredError: number;
  readonly meanEvalMeanSquaredError: number;
  readonly evalMeanSquaredErrorCi95: ConfidenceInterval;
  readonly stdevEvalMeanSquaredError: number;
  readonly meanExactWithinToleranceFraction: number;
  readonly exactWithinToleranceFractionCi95: ConfidenceInterval;
  readonly stdevExactWithinToleranceFraction: number;
  readonly meanDualActiveRegionMeanAbsoluteError: number;
  readonly dualActiveRegionMeanAbsoluteErrorCi95: ConfidenceInterval;
  readonly stdevDualActiveRegionMeanAbsoluteError: number;
  readonly meanCodecRoundTripExactFraction: number;
  readonly meanReassemblyExactFraction: number;
  readonly meanFoldInvarianceExactFraction: number;
  readonly meanFrameCount: number;
  readonly meanPayloadBytes: number;
  readonly samplePredictions: readonly AeonFramedTransformerSamplePrediction[];
}

export interface AeonFramedTransformerTopologyMetrics {
  readonly nodeCount: number;
  readonly edgeCount: number;
  readonly structuralBeta1: number;
  readonly transformerletCount: number;
  readonly rotationStageCount: number;
  readonly parameterCount: number;
}

export interface GnosisAeonFramedTransformerBenchmarkReport {
  readonly label: AeonFramedTransformerConfig['label'];
  readonly config: {
    readonly seedCount: number;
    readonly epochs: number;
    readonly learningRate: number;
    readonly exactPredictionTolerance: number;
    readonly dualActiveThreshold: number;
    readonly frameStageCount: number;
    readonly trainSampleCount: number;
    readonly evalSampleCount: number;
    readonly targetFamily: 'quadrant-l1-framed-transformer';
  };
  readonly topology: AeonFramedTransformerTopologyMetrics;
  readonly strategies: Record<
    AeonFramedTransformerStrategy,
    AeonFramedTransformerStrategyReport
  >;
  readonly rankingByEvalMeanSquaredError: readonly AeonFramedTransformerStrategy[];
  readonly predictedRankingMatches: boolean;
}

interface AeonFramedTransformerSample {
  readonly leftInput: number;
  readonly rightInput: number;
  readonly target: number;
}

interface TransformerletParameters {
  attentionWeight: number;
  attentionBias: number;
  feedForwardWeight: number;
  feedForwardBias: number;
}

interface ParsedTopology {
  readonly strategy: AeonFramedTransformerStrategy;
  readonly path: string;
  readonly metrics: AeonFramedTransformerTopologyMetrics;
}

interface TransformerletSpec {
  readonly id: string;
  readonly streamBaseId: number;
  readonly signedProjection: (sample: AeonFramedTransformerSample) => number;
  readonly residualProjection: (sample: AeonFramedTransformerSample) => number;
}

interface TransformerletActivation {
  readonly attention: number;
  readonly residual: number;
  readonly feedForward: number;
  readonly output: number;
  readonly signedInput: number;
}

interface PredictionResult {
  readonly prediction: number;
  readonly selectedTransformerletIndex: number;
  readonly activations: readonly TransformerletActivation[];
}

interface EvaluationMetrics {
  readonly meanSquaredError: number;
  readonly exactWithinToleranceFraction: number;
  readonly dualActiveRegionMeanAbsoluteError: number;
}

interface TransportMetrics {
  readonly codecRoundTripExactFraction: number;
  readonly reassemblyExactFraction: number;
  readonly foldInvarianceExactFraction: number;
  readonly frameCount: number;
  readonly payloadBytes: number;
}

interface FramedPredictionResult extends TransportMetrics {
  readonly prediction: number;
}

const DEFAULT_AEON_FRAMED_TRANSFORMER_TOPOLOGY_URLS: Record<
  AeonFramedTransformerStrategy,
  URL
> = {
  linear: new URL(
    '../../examples/benchmarks/aeon-framed-transformer-linear.gg',
    import.meta.url
  ),
  'winner-take-all': new URL(
    '../../examples/benchmarks/aeon-framed-transformer-winner-take-all.gg',
    import.meta.url
  ),
  'early-stop': new URL(
    '../../examples/benchmarks/aeon-framed-transformer-early-stop.gg',
    import.meta.url
  ),
};

export const DEFAULT_AEON_FRAMED_TRANSFORMER_TOPOLOGY_FILES =
  Object.fromEntries(
    (
      Object.entries(DEFAULT_AEON_FRAMED_TRANSFORMER_TOPOLOGY_URLS) as Array<
        readonly [AeonFramedTransformerStrategy, URL]
      >
    ).map(([strategy, url]) => [strategy, fileURLToPath(url)])
  ) as Record<AeonFramedTransformerStrategy, string>;

const transformerletSpecs: readonly TransformerletSpec[] = [
  {
    id: 'x_pos',
    streamBaseId: 1,
    signedProjection: (sample) => sample.leftInput,
    residualProjection: (sample) => sample.leftInput,
  },
  {
    id: 'x_neg',
    streamBaseId: 2,
    signedProjection: (sample) => -sample.leftInput,
    residualProjection: (sample) => sample.leftInput,
  },
  {
    id: 'y_pos',
    streamBaseId: 3,
    signedProjection: (sample) => sample.rightInput,
    residualProjection: (sample) => sample.rightInput,
  },
  {
    id: 'y_neg',
    streamBaseId: 4,
    signedProjection: (sample) => -sample.rightInput,
    residualProjection: (sample) => sample.rightInput,
  },
] as const;

function sigmoid(value: number): number {
  return 1 / (1 + Math.exp(-value));
}

function countConnectedComponents(
  nodeIds: readonly string[],
  edges: readonly {
    sourceIds: readonly string[];
    targetIds: readonly string[];
  }[]
): number {
  const neighbors = new Map<string, Set<string>>();
  for (const nodeId of nodeIds) {
    neighbors.set(nodeId, new Set<string>());
  }

  for (const edge of edges) {
    for (const sourceId of edge.sourceIds) {
      const sourceNeighbors = neighbors.get(sourceId);
      if (!sourceNeighbors) {
        continue;
      }
      for (const targetId of edge.targetIds) {
        sourceNeighbors.add(targetId);
        neighbors.get(targetId)?.add(sourceId);
      }
    }
  }

  let components = 0;
  const visited = new Set<string>();
  for (const nodeId of nodeIds) {
    if (visited.has(nodeId)) {
      continue;
    }

    components++;
    const frontier = [nodeId];
    while (frontier.length > 0) {
      const current = frontier.pop();
      if (current === undefined || visited.has(current)) {
        continue;
      }

      visited.add(current);
      for (const neighbor of neighbors.get(current) ?? []) {
        if (!visited.has(neighbor)) {
          frontier.push(neighbor);
        }
      }
    }
  }

  return Math.max(1, components);
}

function structuralBeta1(
  nodeIds: readonly string[],
  edges: readonly {
    sourceIds: readonly string[];
    targetIds: readonly string[];
  }[]
): number {
  const expandedEdgeCount = edges.reduce(
    (sum, edge) => sum + edge.sourceIds.length * edge.targetIds.length,
    0
  );

  return Math.max(
    0,
    expandedEdgeCount -
      nodeIds.length +
      countConnectedComponents(nodeIds, edges)
  );
}

function quadrantRoutingTarget(
  sample: Omit<AeonFramedTransformerSample, 'target'>
): number {
  return Math.abs(sample.leftInput) + Math.abs(sample.rightInput);
}

function buildSamples(axis: readonly number[]): AeonFramedTransformerSample[] {
  return axis.flatMap((leftInput) =>
    axis.map((rightInput) => ({
      leftInput,
      rightInput,
      target: quadrantRoutingTarget({ leftInput, rightInput }),
    }))
  );
}

function initializeTransformerletParameters(
  seed: number
): TransformerletParameters[] {
  const random = createDeterministicRandom(seed);
  return transformerletSpecs.map(() => ({
    attentionWeight: (random() - 0.5) * 1.6,
    attentionBias: (random() - 0.5) * 0.6,
    feedForwardWeight: (random() - 0.5) * 1.0,
    feedForwardBias: (random() - 0.5) * 0.4,
  }));
}

function activateTransformerlets(
  parameters: readonly TransformerletParameters[],
  sample: AeonFramedTransformerSample
): readonly TransformerletActivation[] {
  return transformerletSpecs.map((transformerlet, index) => {
    const parameter = parameters[index];
    if (!parameter) {
      throw new Error(`Missing transformerlet parameters for index ${index}`);
    }

    const signedInput = transformerlet.signedProjection(sample);
    const residual = transformerlet.residualProjection(sample);
    const attention = sigmoid(
      parameter.attentionWeight * signedInput + parameter.attentionBias
    );
    const feedForward =
      parameter.feedForwardWeight * residual + parameter.feedForwardBias;

    return {
      attention,
      residual,
      feedForward,
      output: attention * feedForward,
      signedInput,
    };
  });
}

function predictFromActivations(
  strategy: AeonFramedTransformerStrategy,
  activations: readonly TransformerletActivation[]
): PredictionResult {
  let selectedTransformerletIndex = 0;
  for (let index = 1; index < activations.length; index++) {
    const selected = activations[selectedTransformerletIndex];
    const candidate = activations[index];
    if (
      candidate &&
      selected &&
      Math.abs(candidate.output) > Math.abs(selected.output)
    ) {
      selectedTransformerletIndex = index;
    }
  }

  if (strategy === 'linear') {
    return {
      prediction: activations.reduce(
        (sum, activation) => sum + activation.output,
        0
      ),
      selectedTransformerletIndex,
      activations,
    };
  }

  if (strategy === 'early-stop') {
    return {
      prediction: activations[0]?.output ?? 0,
      selectedTransformerletIndex: 0,
      activations,
    };
  }

  return {
    prediction: activations[selectedTransformerletIndex]?.output ?? 0,
    selectedTransformerletIndex,
    activations,
  };
}

function predict(
  strategy: AeonFramedTransformerStrategy,
  parameters: readonly TransformerletParameters[],
  sample: AeonFramedTransformerSample
): PredictionResult {
  return predictFromActivations(
    strategy,
    activateTransformerlets(parameters, sample)
  );
}

function updateParameters(
  strategy: AeonFramedTransformerStrategy,
  parameters: TransformerletParameters[],
  sample: AeonFramedTransformerSample,
  learningRate: number
): void {
  const result = predict(strategy, parameters, sample);
  const error = result.prediction - sample.target;

  for (let index = 0; index < parameters.length; index++) {
    if (
      strategy === 'winner-take-all' &&
      index !== result.selectedTransformerletIndex
    ) {
      continue;
    }
    if (strategy === 'early-stop' && index !== 0) {
      continue;
    }

    const parameter = parameters[index];
    const activation = result.activations[index];
    if (!parameter || !activation) {
      continue;
    }

    const attentionPrime = activation.attention * (1 - activation.attention);
    parameter.feedForwardWeight -=
      learningRate * error * activation.attention * activation.residual;
    parameter.feedForwardBias -= learningRate * error * activation.attention;
    parameter.attentionWeight -=
      learningRate *
      error *
      attentionPrime *
      activation.signedInput *
      activation.feedForward;
    parameter.attentionBias -=
      learningRate * error * attentionPrime * activation.feedForward;
  }
}

function evaluateSamplesDirect(
  strategy: AeonFramedTransformerStrategy,
  parameters: readonly TransformerletParameters[],
  samples: readonly AeonFramedTransformerSample[],
  exactPredictionTolerance: number,
  dualActiveThreshold: number
): EvaluationMetrics {
  if (samples.length === 0) {
    return {
      meanSquaredError: 0,
      exactWithinToleranceFraction: 0,
      dualActiveRegionMeanAbsoluteError: 0,
    };
  }

  let squaredErrorSum = 0;
  let exactCount = 0;
  let dualActiveAbsoluteErrorSum = 0;
  let dualActiveCount = 0;

  for (const sample of samples) {
    const prediction = predict(strategy, parameters, sample).prediction;
    const error = prediction - sample.target;
    squaredErrorSum += error * error;

    if (Math.abs(error) <= exactPredictionTolerance) {
      exactCount++;
    }

    if (
      Math.abs(sample.leftInput) >= dualActiveThreshold &&
      Math.abs(sample.rightInput) >= dualActiveThreshold
    ) {
      dualActiveAbsoluteErrorSum += Math.abs(error);
      dualActiveCount++;
    }
  }

  return {
    meanSquaredError: squaredErrorSum / samples.length,
    exactWithinToleranceFraction: exactCount / samples.length,
    dualActiveRegionMeanAbsoluteError:
      dualActiveCount === 0 ? 0 : dualActiveAbsoluteErrorSum / dualActiveCount,
  };
}

function encodePayload(values: readonly number[]): Uint8Array {
  const floatValues = new Float64Array(values.length);
  values.forEach((value, index) => {
    floatValues[index] = value;
  });
  return new Uint8Array(floatValues.buffer.slice(0));
}

function decodePayload(payload: Uint8Array): readonly number[] {
  const aligned = payload.slice();
  return Array.from(
    new Float64Array(aligned.buffer, aligned.byteOffset, aligned.byteLength / 8)
  );
}

function framesEqual(left: FlowFrame, right: FlowFrame): boolean {
  if (
    left.streamId !== right.streamId ||
    left.sequence !== right.sequence ||
    left.flags !== right.flags ||
    left.payload.length !== right.payload.length
  ) {
    return false;
  }

  for (let index = 0; index < left.payload.length; index++) {
    if (left.payload[index] !== right.payload[index]) {
      return false;
    }
  }

  return true;
}

function buildFlowFrames(
  sampleIndex: number,
  activations: readonly TransformerletActivation[]
): readonly FlowFrame[] {
  const streamOffset = sampleIndex * 8;

  return activations.flatMap((activation, index) => {
    const transformerlet = transformerletSpecs[index];
    if (!transformerlet) {
      throw new Error(`Missing transformerlet spec for index ${index}`);
    }

    const streamId = streamOffset + transformerlet.streamBaseId;
    return [
      {
        streamId,
        sequence: 0,
        flags: FORK,
        payload: encodePayload([activation.signedInput, activation.residual]),
      },
      {
        streamId,
        sequence: 1,
        flags: RACE,
        payload: encodePayload([
          activation.attention,
          activation.attention * activation.residual,
        ]),
      },
      {
        streamId,
        sequence: 2,
        flags: RACE,
        payload: encodePayload([
          activation.feedForward,
          activation.feedForward * activation.residual,
        ]),
      },
      {
        streamId,
        sequence: 3,
        flags: FOLD | FIN,
        payload: encodePayload([activation.output]),
      },
    ] satisfies readonly FlowFrame[];
  });
}

function reconstructPrediction(
  strategy: AeonFramedTransformerStrategy,
  frames: readonly FlowFrame[]
): number {
  const outputs = transformerletSpecs.map((transformerlet) => {
    const streamFrames = frames
      .filter((frame) => frame.streamId % 8 === transformerlet.streamBaseId % 8)
      .sort((left, right) => left.sequence - right.sequence);
    const collapseFrame = streamFrames.find((frame) => frame.sequence === 3);
    if (!collapseFrame) {
      throw new Error(
        `Missing collapse frame for stream base ${transformerlet.streamBaseId}`
      );
    }
    const [output] = decodePayload(collapseFrame.payload);
    return output ?? 0;
  });

  if (strategy === 'linear') {
    return outputs.reduce((sum, value) => sum + value, 0);
  }
  if (strategy === 'early-stop') {
    return outputs[0] ?? 0;
  }

  return outputs.reduce((selected, candidate) =>
    Math.abs(candidate) > Math.abs(selected) ? candidate : selected
  );
}

function frameSort(left: FlowFrame, right: FlowFrame): number {
  return left.streamId - right.streamId || left.sequence - right.sequence;
}

function framedPredict(
  strategy: AeonFramedTransformerStrategy,
  parameters: readonly TransformerletParameters[],
  sample: AeonFramedTransformerSample,
  sampleIndex: number
): FramedPredictionResult {
  const direct = predict(strategy, parameters, sample);
  const originalFrames = buildFlowFrames(sampleIndex, direct.activations);
  const codec = FlowCodec.createSync();

  let codecExactCount = 0;
  const decodedFrames = originalFrames.map((frame) => {
    const encoded = codec.encode(frame);
    const decoded = codec.decode(encoded).frame;
    if (framesEqual(frame, decoded)) {
      codecExactCount++;
    }
    return decoded;
  });

  const reassembler = new FrameReassembler();
  const deliveredFrames: FlowFrame[] = [];
  for (const frame of [...decodedFrames].reverse()) {
    deliveredFrames.push(...reassembler.push(frame));
  }

  const sortedOriginal = [...originalFrames].sort(frameSort);
  const sortedDelivered = [...deliveredFrames].sort(frameSort);
  let reassemblyExactCount = 0;
  for (let index = 0; index < sortedOriginal.length; index++) {
    const originalFrame = sortedOriginal[index];
    const deliveredFrame = sortedDelivered[index];
    if (
      originalFrame &&
      deliveredFrame &&
      framesEqual(originalFrame, deliveredFrame)
    ) {
      reassemblyExactCount++;
    }
  }

  const reconstructedPrediction = reconstructPrediction(
    strategy,
    sortedDelivered
  );
  const payloadBytes = originalFrames.reduce(
    (sum, frame) => sum + frame.payload.length,
    0
  );

  return {
    prediction: direct.prediction,
    codecRoundTripExactFraction:
      originalFrames.length === 0 ? 0 : codecExactCount / originalFrames.length,
    reassemblyExactFraction:
      originalFrames.length === 0
        ? 0
        : reassemblyExactCount / originalFrames.length,
    foldInvarianceExactFraction:
      reconstructedPrediction === direct.prediction ? 1 : 0,
    frameCount: originalFrames.length,
    payloadBytes,
  };
}

function evaluateSamplesWithTransport(
  strategy: AeonFramedTransformerStrategy,
  parameters: readonly TransformerletParameters[],
  samples: readonly AeonFramedTransformerSample[],
  exactPredictionTolerance: number,
  dualActiveThreshold: number
): EvaluationMetrics & TransportMetrics {
  if (samples.length === 0) {
    return {
      meanSquaredError: 0,
      exactWithinToleranceFraction: 0,
      dualActiveRegionMeanAbsoluteError: 0,
      codecRoundTripExactFraction: 0,
      reassemblyExactFraction: 0,
      foldInvarianceExactFraction: 0,
      frameCount: 0,
      payloadBytes: 0,
    };
  }

  let squaredErrorSum = 0;
  let exactCount = 0;
  let dualActiveAbsoluteErrorSum = 0;
  let dualActiveCount = 0;
  let codecRoundTripExactFractionSum = 0;
  let reassemblyExactFractionSum = 0;
  let foldInvarianceExactFractionSum = 0;
  let frameCountSum = 0;
  let payloadBytesSum = 0;

  for (const [sampleIndex, sample] of samples.entries()) {
    const framed = framedPredict(strategy, parameters, sample, sampleIndex);
    const error = framed.prediction - sample.target;
    squaredErrorSum += error * error;

    if (Math.abs(error) <= exactPredictionTolerance) {
      exactCount++;
    }

    if (
      Math.abs(sample.leftInput) >= dualActiveThreshold &&
      Math.abs(sample.rightInput) >= dualActiveThreshold
    ) {
      dualActiveAbsoluteErrorSum += Math.abs(error);
      dualActiveCount++;
    }

    codecRoundTripExactFractionSum += framed.codecRoundTripExactFraction;
    reassemblyExactFractionSum += framed.reassemblyExactFraction;
    foldInvarianceExactFractionSum += framed.foldInvarianceExactFraction;
    frameCountSum += framed.frameCount;
    payloadBytesSum += framed.payloadBytes;
  }

  return {
    meanSquaredError: squaredErrorSum / samples.length,
    exactWithinToleranceFraction: exactCount / samples.length,
    dualActiveRegionMeanAbsoluteError:
      dualActiveCount === 0 ? 0 : dualActiveAbsoluteErrorSum / dualActiveCount,
    codecRoundTripExactFraction:
      codecRoundTripExactFractionSum / samples.length,
    reassemblyExactFraction: reassemblyExactFractionSum / samples.length,
    foldInvarianceExactFraction:
      foldInvarianceExactFractionSum / samples.length,
    frameCount: frameCountSum / samples.length,
    payloadBytes: payloadBytesSum / samples.length,
  };
}

function summarizeSeedMetrics(
  metrics: readonly AeonFramedTransformerSeedMetrics[]
): Omit<
  AeonFramedTransformerStrategyReport,
  | 'topologyPath'
  | 'topologyNodeCount'
  | 'topologyEdgeCount'
  | 'structuralBeta1'
  | 'transformerletCount'
  | 'rotationStageCount'
  | 'parameterCount'
  | 'streamCountPerSample'
  | 'samplePredictions'
> {
  const evalMeanSquaredErrors = metrics.map(
    (entry) => entry.evalMeanSquaredError
  );
  const exactFractions = metrics.map(
    (entry) => entry.exactWithinToleranceFraction
  );
  const dualActiveErrors = metrics.map(
    (entry) => entry.dualActiveRegionMeanAbsoluteError
  );

  return {
    seedMetrics: metrics,
    meanTrainMeanSquaredError: mean(
      metrics.map((entry) => entry.trainMeanSquaredError)
    ),
    stdevTrainMeanSquaredError: stdev(
      metrics.map((entry) => entry.trainMeanSquaredError)
    ),
    meanEvalMeanSquaredError: mean(evalMeanSquaredErrors),
    evalMeanSquaredErrorCi95: bootstrapMeanConfidenceInterval(
      evalMeanSquaredErrors,
      {
        seed: 0x41454f4e,
      }
    ),
    stdevEvalMeanSquaredError: stdev(evalMeanSquaredErrors),
    meanExactWithinToleranceFraction: mean(exactFractions),
    exactWithinToleranceFractionCi95: bootstrapMeanConfidenceInterval(
      exactFractions,
      {
        seed: 0x5452414e,
      }
    ),
    stdevExactWithinToleranceFraction: stdev(exactFractions),
    meanDualActiveRegionMeanAbsoluteError: mean(dualActiveErrors),
    dualActiveRegionMeanAbsoluteErrorCi95: bootstrapMeanConfidenceInterval(
      dualActiveErrors,
      {
        seed: 0x53464f4c,
      }
    ),
    stdevDualActiveRegionMeanAbsoluteError: stdev(dualActiveErrors),
    meanCodecRoundTripExactFraction: mean(
      metrics.map((entry) => entry.codecRoundTripExactFraction)
    ),
    meanReassemblyExactFraction: mean(
      metrics.map((entry) => entry.reassemblyExactFraction)
    ),
    meanFoldInvarianceExactFraction: mean(
      metrics.map((entry) => entry.foldInvarianceExactFraction)
    ),
    meanFrameCount: mean(metrics.map((entry) => entry.frameCount)),
    meanPayloadBytes: mean(metrics.map((entry) => entry.payloadBytes)),
  };
}

function samplePredictions(
  strategy: AeonFramedTransformerStrategy,
  trainedParameterSets: readonly (readonly TransformerletParameters[])[],
  samplePoints: readonly (readonly [number, number])[]
): AeonFramedTransformerSamplePrediction[] {
  return samplePoints.map(([leftInput, rightInput]) => {
    const sample: AeonFramedTransformerSample = {
      leftInput,
      rightInput,
      target: quadrantRoutingTarget({ leftInput, rightInput }),
    };
    const predictions = trainedParameterSets.map(
      (parameterSet, seedIndex) =>
        framedPredict(strategy, parameterSet, sample, seedIndex).prediction
    );
    const frameCounts = trainedParameterSets.map(
      (parameterSet, seedIndex) =>
        framedPredict(strategy, parameterSet, sample, seedIndex).frameCount
    );
    const payloadBytes = trainedParameterSets.map(
      (parameterSet, seedIndex) =>
        framedPredict(strategy, parameterSet, sample, seedIndex).payloadBytes
    );

    return {
      leftInput,
      rightInput,
      target: sample.target,
      meanPrediction: mean(predictions),
      predictionStddev: stdev(predictions),
      meanFrameCount: mean(frameCounts),
      meanPayloadBytes: mean(payloadBytes),
    };
  });
}

/**
 * Handles the gnosis make Default Aeon Framed Transformer Config workflow.
 */
export function makeDefaultAeonFramedTransformerConfig(): AeonFramedTransformerConfig {
  return {
    label: 'gnosis-aeon-framed-transformer-benchmark-v1',
    seedCount: 12,
    epochs: 1800,
    learningRate: 0.02,
    exactPredictionTolerance: 0.08,
    dualActiveThreshold: 0.4,
    frameStageCount: 4,
    trainAxis: [
      -1.5, -1.25, -1.0, -0.75, -0.5, -0.25, 0, 0.25, 0.5, 0.75, 1.0, 1.25, 1.5,
    ],
    evalAxis: [-1.4, -1.1, -0.8, -0.4, -0.1, 0.1, 0.4, 0.8, 1.1, 1.4],
    samplePoints: [
      [-1.2, -1.2],
      [-1.2, 1.2],
      [1.2, -1.2],
      [1.2, 1.2],
      [1.2, 0],
    ],
  };
}

async function loadTopology(
  strategy: AeonFramedTransformerStrategy
): Promise<ParsedTopology> {
  const path = DEFAULT_AEON_FRAMED_TRANSFORMER_TOPOLOGY_FILES[strategy];
  const source = await readFile(path, 'utf8');
  const program = parseGgProgram(lowerUfcsSource(source));
  const foldEdges = program.edges.filter(
    (edge) => edge.type.toUpperCase() === 'FOLD'
  );
  const benchmarkFoldEdges = foldEdges.filter(
    (edge) => edge.properties.strategy === strategy
  );
  if (benchmarkFoldEdges.length !== 1) {
    throw new Error(
      `Expected exactly one benchmark FOLD edge with strategy "${strategy}" in ${path}`
    );
  }

  const foldStrategy = benchmarkFoldEdges[0]?.properties.strategy;
  if (foldStrategy !== strategy) {
    throw new Error(
      `Expected FOLD strategy "${strategy}" in ${path}, found "${
        foldStrategy ?? 'missing'
      }"`
    );
  }

  const transformerletNodes = program.nodes.filter((node) =>
    node.labels.includes('ToyTransformerlet')
  );
  if (transformerletNodes.length !== 4) {
    throw new Error(`Expected exactly four ToyTransformerlet nodes in ${path}`);
  }

  const rotationNodes = program.nodes.filter((node) =>
    node.labels.includes('RotationScheduler')
  );
  if (rotationNodes.length === 0) {
    throw new Error(`Expected at least one RotationScheduler node in ${path}`);
  }

  const parameterCount = transformerletNodes.reduce((sum, node) => {
    const rawParameters = Number(node.properties.parameters ?? 0);
    return sum + (Number.isFinite(rawParameters) ? rawParameters : 0);
  }, 0);

  const rotationStageCount = rotationNodes.reduce((maxStages, node) => {
    const rawStages = Number(node.properties.stages ?? 0);
    return Number.isFinite(rawStages)
      ? Math.max(maxStages, rawStages)
      : maxStages;
  }, 0);

  return {
    strategy,
    path,
    metrics: {
      nodeCount: program.nodes.length,
      edgeCount: program.edges.length,
      structuralBeta1: structuralBeta1(
        program.nodes.map((node) => node.id),
        program.edges.map((edge) => ({
          sourceIds: edge.sourceIds,
          targetIds: edge.targetIds,
        }))
      ),
      transformerletCount: transformerletNodes.length,
      rotationStageCount: Number.isFinite(rotationStageCount)
        ? rotationStageCount
        : 0,
      parameterCount,
    },
  };
}

async function loadDefaultTopologySuite(): Promise<
  Record<AeonFramedTransformerStrategy, ParsedTopology>
> {
  const entries = await Promise.all(
    (
      Object.keys(
        DEFAULT_AEON_FRAMED_TRANSFORMER_TOPOLOGY_FILES
      ) as AeonFramedTransformerStrategy[]
    ).map(async (strategy) => [strategy, await loadTopology(strategy)] as const)
  );

  return Object.fromEntries(entries) as Record<
    AeonFramedTransformerStrategy,
    ParsedTopology
  >;
}

function assertSharedTopology(
  topologies: Record<AeonFramedTransformerStrategy, ParsedTopology>
): AeonFramedTransformerTopologyMetrics {
  const baseline = topologies.linear.metrics;
  for (const strategy of Object.keys(
    topologies
  ) as AeonFramedTransformerStrategy[]) {
    const metrics = topologies[strategy].metrics;
    const isShared =
      metrics.nodeCount === baseline.nodeCount &&
      metrics.edgeCount === baseline.edgeCount &&
      metrics.structuralBeta1 === baseline.structuralBeta1 &&
      metrics.transformerletCount === baseline.transformerletCount &&
      metrics.rotationStageCount === baseline.rotationStageCount &&
      metrics.parameterCount === baseline.parameterCount;
    if (!isShared) {
      throw new Error(
        `Topology mismatch for ${strategy}; benchmark is not parameter-matched`
      );
    }
  }

  return baseline;
}

function trainStrategy(
  strategy: AeonFramedTransformerStrategy,
  config: AeonFramedTransformerConfig,
  trainSamples: readonly AeonFramedTransformerSample[],
  evalSamples: readonly AeonFramedTransformerSample[]
): {
  readonly seedMetrics: readonly AeonFramedTransformerSeedMetrics[];
  readonly trainedParameters: readonly (readonly TransformerletParameters[])[];
} {
  const seedMetrics: AeonFramedTransformerSeedMetrics[] = [];
  const trainedParameters: TransformerletParameters[][] = [];

  for (let seedIndex = 0; seedIndex < config.seedCount; seedIndex++) {
    const seed = seedIndex + 1;
    const parameters = initializeTransformerletParameters(seed);
    const random = createDeterministicRandom((seed * 0x9e3779b1) >>> 0);

    for (let epoch = 0; epoch < config.epochs; epoch++) {
      const parityOffset = Math.floor(random() * trainSamples.length);
      for (
        let sampleOffset = 0;
        sampleOffset < trainSamples.length;
        sampleOffset++
      ) {
        const sample =
          trainSamples[(sampleOffset + parityOffset) % trainSamples.length];
        if (sample) {
          updateParameters(strategy, parameters, sample, config.learningRate);
        }
      }
    }

    const trainEvaluation = evaluateSamplesDirect(
      strategy,
      parameters,
      trainSamples,
      config.exactPredictionTolerance,
      config.dualActiveThreshold
    );
    const evalEvaluation = evaluateSamplesWithTransport(
      strategy,
      parameters,
      evalSamples,
      config.exactPredictionTolerance,
      config.dualActiveThreshold
    );

    seedMetrics.push({
      seed,
      trainMeanSquaredError: trainEvaluation.meanSquaredError,
      evalMeanSquaredError: evalEvaluation.meanSquaredError,
      exactWithinToleranceFraction: evalEvaluation.exactWithinToleranceFraction,
      dualActiveRegionMeanAbsoluteError:
        evalEvaluation.dualActiveRegionMeanAbsoluteError,
      codecRoundTripExactFraction: evalEvaluation.codecRoundTripExactFraction,
      reassemblyExactFraction: evalEvaluation.reassemblyExactFraction,
      foldInvarianceExactFraction: evalEvaluation.foldInvarianceExactFraction,
      frameCount: evalEvaluation.frameCount,
      payloadBytes: evalEvaluation.payloadBytes,
    });
    trainedParameters.push(parameters.map((parameter) => ({ ...parameter })));
  }

  return {
    seedMetrics,
    trainedParameters,
  };
}

/**
 * Handles the gnosis run Gnosis Aeon Framed Transformer Benchmark workflow.
 */
export async function runGnosisAeonFramedTransformerBenchmark(
  config: AeonFramedTransformerConfig = makeDefaultAeonFramedTransformerConfig()
): Promise<GnosisAeonFramedTransformerBenchmarkReport> {
  const topologies = await loadDefaultTopologySuite();
  const sharedTopology = assertSharedTopology(topologies);
  const trainSamples = buildSamples(config.trainAxis);
  const evalSamples = buildSamples(config.evalAxis);

  const strategies = {} as Record<
    AeonFramedTransformerStrategy,
    AeonFramedTransformerStrategyReport
  >;
  for (const strategy of Object.keys(
    topologies
  ) as AeonFramedTransformerStrategy[]) {
    const trained = trainStrategy(strategy, config, trainSamples, evalSamples);
    const summary = summarizeSeedMetrics(trained.seedMetrics);

    strategies[strategy] = {
      topologyPath: topologies[strategy].path,
      topologyNodeCount: topologies[strategy].metrics.nodeCount,
      topologyEdgeCount: topologies[strategy].metrics.edgeCount,
      structuralBeta1: topologies[strategy].metrics.structuralBeta1,
      transformerletCount: topologies[strategy].metrics.transformerletCount,
      rotationStageCount: topologies[strategy].metrics.rotationStageCount,
      parameterCount: topologies[strategy].metrics.parameterCount,
      streamCountPerSample: transformerletSpecs.length,
      ...summary,
      samplePredictions: samplePredictions(
        strategy,
        trained.trainedParameters,
        config.samplePoints
      ),
    };
  }

  const rankingByEvalMeanSquaredError = (
    Object.keys(strategies) as AeonFramedTransformerStrategy[]
  ).sort(
    (left, right) =>
      strategies[left].meanEvalMeanSquaredError -
      strategies[right].meanEvalMeanSquaredError
  );

  return {
    label: config.label,
    config: {
      seedCount: config.seedCount,
      epochs: config.epochs,
      learningRate: config.learningRate,
      exactPredictionTolerance: config.exactPredictionTolerance,
      dualActiveThreshold: config.dualActiveThreshold,
      frameStageCount: config.frameStageCount,
      trainSampleCount: trainSamples.length,
      evalSampleCount: evalSamples.length,
      targetFamily: 'quadrant-l1-framed-transformer',
    },
    topology: sharedTopology,
    strategies,
    rankingByEvalMeanSquaredError,
    predictedRankingMatches:
      rankingByEvalMeanSquaredError[0] === 'linear' &&
      rankingByEvalMeanSquaredError[1] === 'winner-take-all' &&
      rankingByEvalMeanSquaredError[2] === 'early-stop',
  };
}

function formatSamplePair(
  sample: AeonFramedTransformerSamplePrediction
): string {
  return `(${sample.leftInput.toFixed(2)}, ${sample.rightInput.toFixed(2)})`;
}

function formatInterval(interval: ConfidenceInterval): string {
  return `[${interval.low.toFixed(3)}, ${interval.high.toFixed(3)}]`;
}

/**
 * Handles the gnosis render Gnosis Aeon Framed Transformer Benchmark Markdown workflow.
 */
export function renderGnosisAeonFramedTransformerBenchmarkMarkdown(
  report: GnosisAeonFramedTransformerBenchmarkReport
): string {
  const lines: string[] = [];
  lines.push('# Gnosis Aeon-Framed Transformer Benchmark');
  lines.push('');
  lines.push(`- Label: \`${report.label}\``);
  lines.push(`- Target family: \`${report.config.targetFamily}\``);
  lines.push(`- Seeds: \`${report.config.seedCount}\``);
  lines.push(`- Epochs: \`${report.config.epochs}\``);
  lines.push(`- Learning rate: \`${report.config.learningRate}\``);
  lines.push(`- Frame stages per stream: \`${report.config.frameStageCount}\``);
  lines.push(`- Train samples: \`${report.config.trainSampleCount}\``);
  lines.push(`- Eval samples: \`${report.config.evalSampleCount}\``);
  lines.push(`- Shared parameter count: \`${report.topology.parameterCount}\``);
  lines.push(`- Shared structural β₁: \`${report.topology.structuralBeta1}\``);
  lines.push(
    `- Rotation stage count: \`${report.topology.rotationStageCount}\``
  );
  lines.push(
    `- Predicted ranking recovered: \`${
      report.predictedRankingMatches ? 'yes' : 'no'
    }\``
  );
  lines.push('');
  lines.push('## Aggregated Metrics');
  lines.push('');
  lines.push(
    '| Strategy | Mean eval MSE | Eval MSE 95% CI | Exact-within-tolerance | Exact 95% CI | Dual-active abs error | Codec round-trip | Reassembly | Fold invariance | Mean frames/sample | Mean payload bytes/sample |'
  );
  lines.push('|---|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|');

  for (const strategy of report.rankingByEvalMeanSquaredError) {
    const metrics = report.strategies[strategy];
    lines.push(
      `| \`${strategy}\` | ${metrics.meanEvalMeanSquaredError.toFixed(
        3
      )} | ${formatInterval(
        metrics.evalMeanSquaredErrorCi95
      )} | ${metrics.meanExactWithinToleranceFraction.toFixed(
        3
      )} | ${formatInterval(
        metrics.exactWithinToleranceFractionCi95
      )} | ${metrics.meanDualActiveRegionMeanAbsoluteError.toFixed(
        3
      )} | ${metrics.meanCodecRoundTripExactFraction.toFixed(
        3
      )} | ${metrics.meanReassemblyExactFraction.toFixed(
        3
      )} | ${metrics.meanFoldInvarianceExactFraction.toFixed(
        3
      )} | ${metrics.meanFrameCount.toFixed(
        1
      )} | ${metrics.meanPayloadBytes.toFixed(1)} |`
    );
  }

  lines.push('');
  lines.push('## Sample Predictions');
  lines.push('');
  lines.push(
    '| Strategy | Input pair | Target | Mean prediction | Prediction σ | Mean frames | Mean payload bytes |'
  );
  lines.push('|---|---|---:|---:|---:|---:|---:|');

  for (const strategy of report.rankingByEvalMeanSquaredError) {
    for (const sample of report.strategies[strategy].samplePredictions) {
      lines.push(
        `| \`${strategy}\` | ${formatSamplePair(
          sample
        )} | ${sample.target.toFixed(3)} | ${sample.meanPrediction.toFixed(
          3
        )} | ${sample.predictionStddev.toFixed(
          3
        )} | ${sample.meanFrameCount.toFixed(
          1
        )} | ${sample.meanPayloadBytes.toFixed(1)} |`
      );
    }
  }

  lines.push('');
  lines.push(
    'Interpretation: a Wallington-style staged input boundary feeds four toy transformerlets that each emit real Aeon flow frames, while the Worthington-style collapse boundary varies only the fold law. Linear collapse alone can aggregate the two simultaneously active transformerlets required for the `|x| + |y|` target, and the same framed outputs survive codec round-trips plus out-of-order reassembly without changing the result.'
  );
  lines.push('');
  return `${lines.join('\n')}\n`;
}
