// @a0n/gnosis/src/benchmarks/bitwise-variants-benchmark.ts
/**
 * bitwise-variants-benchmark.ts -- Bitwise Cipher Variants benchmark suite
 *
 * Measures performance of the three variants:
 * 1. Static Helix Cipher
 * 2. Rotating Helix Cipher
 * 3. Lacey Cipher (Rotating + AEAD)
 *
 * Usage: gnode run src/benchmarks/bitwise-variants-benchmark.ts
 */

import {
  helixKeyFromSeed,
  helixEncryptFrame,
  helixDecryptFrame,
  rotatingHelixKeyFromSeed,
  rotatingHelixEncryptFrame,
  rotatingHelixDecryptFrame,
  laceyEncryptFrame,
  laceyDecryptFrame,
} from '../bitwise-cipher';

// ============================================================================
// Benchmark harness
// ============================================================================

interface BenchResult {
  name: string;
  ops: number;
  nsPerOp: number;
  opsPerSec: number;
}

function benchSync(
  name: string,
  fn: () => void,
  warmup = 1000,
  iterations = 100000
): BenchResult {
  for (let i = 0; i < warmup; i++) fn();

  const start = performance.now();
  for (let i = 0; i < iterations; i++) fn();
  const elapsed = performance.now() - start;

  const nsPerOp = (elapsed * 1e6) / iterations;
  const opsPerSec = iterations / (elapsed / 1000);

  return { name, ops: iterations, nsPerOp, opsPerSec };
}

async function benchAsync(
  name: string,
  fn: () => Promise<void>,
  warmup = 100,
  iterations = 10000
): Promise<BenchResult> {
  for (let i = 0; i < warmup; i++) await fn();

  const start = performance.now();
  for (let i = 0; i < iterations; i++) await fn();
  const elapsed = performance.now() - start;

  const nsPerOp = (elapsed * 1e6) / iterations;
  const opsPerSec = iterations / (elapsed / 1000);

  return { name, ops: iterations, nsPerOp, opsPerSec };
}

function report(results: BenchResult[]): void {
  console.log(
    '\n╔══════════════════════════════════════════════════════════════╗'
  );
  console.log(
    '║           Bitwise Cipher Variants Benchmarks                ║'
  );
  console.log(
    '╠══════════════════════════════════════════════════════════════╣'
  );
  for (const r of results) {
    const name = r.name.padEnd(35);
    const ns = r.nsPerOp.toFixed(1).padStart(10);
    const ops = formatNumber(r.opsPerSec).padStart(12);
    console.log(`║ ${name} ${ns} ns/op  ${ops} ops/s ║`);
  }
  console.log(
    '╚══════════════════════════════════════════════════════════════╝'
  );
}

function formatNumber(n: number): string {
  if (n >= 1e9) return (n / 1e9).toFixed(2) + 'B';
  if (n >= 1e6) return (n / 1e6).toFixed(2) + 'M';
  if (n >= 1e3) return (n / 1e3).toFixed(2) + 'K';
  return n.toFixed(0);
}

// ============================================================================
// Scenarios
// ============================================================================

/**
 * Runs the gnosis command-line workflow.
 */
export async function main() {
  const results: BenchResult[] = [];
  
  // Setup Keys
  const seed = crypto.getRandomValues(new Uint8Array(64));
  const staticKey = helixKeyFromSeed(seed);
  const rotatingKey = rotatingHelixKeyFromSeed(seed);
  const keyPhrase = Array.from(seed).map(b => b.toString(16).padStart(2, '0')).join('');

  // Setup Test Data
  const framePayload = new Uint8Array(256);
  crypto.getRandomValues(framePayload);
  
  // A standard FlowFrame header is 10 bytes:
  // [0..1] stream_id
  // [2..5] position (u32)
  // [6] reserved
  // [7..9] payload_length (u24)
  const plainFrame = new Uint8Array(10 + framePayload.length);
  // fake header
  plainFrame[0] = 0x12; 
  plainFrame[1] = 0x34; // stream id
  plainFrame[2] = 0x00; plainFrame[3] = 0x00; plainFrame[4] = 0x01; plainFrame[5] = 0x00; // pos
  // length (u24)
  plainFrame[7] = (framePayload.length >> 16) & 0xFF;
  plainFrame[8] = (framePayload.length >> 8) & 0xFF;
  plainFrame[9] = framePayload.length & 0xFF;
  // copy payload
  plainFrame.set(framePayload, 10);

  // 1. Static Helix
  const encryptedStatic = helixEncryptFrame(plainFrame, staticKey);
  results.push(benchSync('Static Helix Encrypt', () => {
    helixEncryptFrame(plainFrame, staticKey);
  }));
  results.push(benchSync('Static Helix Decrypt', () => {
    helixDecryptFrame(encryptedStatic, staticKey);
  }));

  // 2. Rotating Helix 
  const encryptedRotating = rotatingHelixEncryptFrame(plainFrame, rotatingKey);
  results.push(benchSync('Rotating Helix Encrypt', () => {
    rotatingHelixEncryptFrame(plainFrame, rotatingKey);
  }));
  results.push(benchSync('Rotating Helix Decrypt', () => {
    rotatingHelixDecryptFrame(encryptedRotating, rotatingKey);
  }));

  // 3. Lacey (Rotating + ChaCha20-Poly1305 AEAD WASM)
  const encryptedLacey = await laceyEncryptFrame(plainFrame, rotatingKey, keyPhrase);
  results.push(await benchAsync('Lacey Cipher Encrypt', async () => {
    await laceyEncryptFrame(plainFrame, rotatingKey, keyPhrase);
  }, 100, 10000));
  
  results.push(await benchAsync('Lacey Cipher Decrypt', async () => {
    await laceyDecryptFrame(encryptedLacey, rotatingKey, keyPhrase, plainFrame.length);
  }, 100, 10000));

  report(results);
}
