// @a0n/gnosis/src/benchmarks/braid-fast-path-benchmark.ts
import { performance } from 'node:perf_hooks';
import { pathToFileURL } from 'node:url';
import {
  createBraidPacketCertificate,
  createBraidFoldCertificate,
  createDecompositionTripleCertificate,
  createExtendedAeonTransitionCertificate,
  createGnosisUringInterferePortCertificate,
  createInterferenceStrideCertificate,
  createPressureWeightedInterferenceStrideCertificate,
  selectCertifiedBraidFastPath,
  type CertifiedBraidFastPathDecision,
} from '../mesh-policy';
import {
  compileTypeScriptToGnosis,
  executeTypeScriptWithGnosis,
  type GnosisTypeScriptBridgeBindings,
  type GnosisTypeScriptBridgeResult,
} from '../ts-bridge';
import {
  bootstrapMeanConfidenceInterval,
  mean,
  stdev,
  type ConfidenceInterval,
} from './statistics';

const BENCHMARK_SOURCE = `
  export function decompose(input: number) {
    return input + 7;
  }

  export function raceAnchor(input: number) {
    return input * 17;
  }

  export function foldAnchor(left: number, right: number) {
    return left + right + 22;
  }

  export async function app(input: { value: number }) {
    const [left, right] = await Promise.all([
      decompose(input.value),
      raceAnchor(input.value),
    ]);
    const folded = foldAnchor(left, right);
    return folded;
  }
`;

const BENCHMARK_BINDINGS: GnosisTypeScriptBridgeBindings = {
  decompose: (input: number) => input + 7,
  raceAnchor: (input: number) => input * 17,
  foldAnchor: (left: number, right: number) => left + right + 22,
};

interface NormalizedBenchmarkOptions {
  readonly iterations: number;
  readonly warmup: number;
}

export interface BraidFastPathBenchmarkOptions {
  readonly iterations?: number;
  readonly warmup?: number;
}

export interface BraidFastPathVariantReport {
  readonly name: 'certified-braid-fast-path' | 'linear-fallback';
  readonly status: 'selected' | 'fallback';
  readonly reason: string;
  readonly iterations: number;
  readonly timingsMs: readonly number[];
  readonly meanMs: number;
  readonly stdevMs: number;
  readonly ci95: ConfidenceInterval;
  readonly totalRuntimeDecisions: number;
  readonly totalAvoidedRuntimeDecisions: number;
  readonly semanticMismatches: number;
  readonly heapDeltaBytes: number;
}

export interface BraidFastPathBenchmarkReport {
  readonly label: 'gnosis-braid-fast-path-benchmark-v1';
  readonly iterations: number;
  readonly warmup: number;
  readonly scheduleKinds: readonly string[];
  readonly runtimeBindingNames: readonly string[];
  readonly runtimeModuleMetadata: {
    readonly emitted: boolean;
    readonly braidFoldDirective: boolean;
    readonly decompositionTripleDirective: boolean;
    readonly braidPacketDirective: boolean;
    readonly extendedAeonDirective: boolean;
    readonly interferePortDirective: boolean;
    readonly pressureWeightedInterferenceDirective: boolean;
    readonly braidFoldCertified: boolean;
    readonly decompositionTripleCertified: boolean;
    readonly braidPacketCertified: boolean;
    readonly extendedAeonCertified: boolean;
    readonly interferePortCertified: boolean;
    readonly pressureWeightedInterferenceCertified: boolean;
  };
  readonly selected: BraidFastPathVariantReport;
  readonly fallback: BraidFastPathVariantReport;
  readonly comparison: {
    readonly meanSpeedupRatio: number;
    readonly meanDeltaMs: number;
    readonly runtimeDecisionDelta: number;
    readonly avoidedRuntimeDecisionDelta: number;
    readonly semanticMismatches: number;
  };
}

interface BraidFastPathVariantAccumulator {
  readonly name: BraidFastPathVariantReport['name'];
  readonly decision: CertifiedBraidFastPathDecision;
  readonly timingsMs: number[];
  status: BraidFastPathVariantReport['status'];
  reason: string;
  totalRuntimeDecisions: number;
  totalAvoidedRuntimeDecisions: number;
  semanticMismatches: number;
  heapDeltaBytes: number;
}

function normalizeBenchmarkOptions(
  options: BraidFastPathBenchmarkOptions = {}
): NormalizedBenchmarkOptions {
  return {
    iterations: Math.max(1, Math.floor(options.iterations ?? 100)),
    warmup: Math.max(0, Math.floor(options.warmup ?? 10)),
  };
}

function parseNumericFlag(
  argv: readonly string[],
  flagName: string,
  fallback: number
): number {
  const inlinePrefix = `--${flagName}=`;
  const inlineValue = argv.find((value) => value.startsWith(inlinePrefix));
  if (inlineValue) {
    const parsed = Number(inlineValue.slice(inlinePrefix.length));
    return Number.isFinite(parsed) ? parsed : fallback;
  }

  const flagIndex = argv.indexOf(`--${flagName}`);
  const value = flagIndex === -1 ? undefined : argv[flagIndex + 1];
  const parsed = value === undefined ? Number.NaN : Number(value);
  return Number.isFinite(parsed) ? parsed : fallback;
}

function expectedPayload(inputValue: number): number {
  return inputValue + 7 + inputValue * 17 + 22;
}

function compileBenchmarkFixture(): GnosisTypeScriptBridgeResult {
  return compileTypeScriptToGnosis(BENCHMARK_SOURCE, {
    exportName: 'app',
    sourceFilePath: 'braid-fast-path-benchmark-input.ts',
  });
}

function createSelectedDecision(): CertifiedBraidFastPathDecision {
  const certificate = createBraidFoldCertificate({
    cycle: 0,
    gateCoordinates: [10, 12, 17, 22],
    saturationTrace: [10, 30, 90],
    closedArchitecture: true,
    endpointsLinked: true,
    baseReady: true,
  });
  return selectCertifiedBraidFastPath(certificate);
}

function createVariantAccumulator(
  name: BraidFastPathVariantReport['name'],
  decision: CertifiedBraidFastPathDecision
): BraidFastPathVariantAccumulator {
  return {
    name,
    decision,
    timingsMs: [],
    status: decision.selected ? 'selected' : 'fallback',
    reason: decision.reason,
    totalRuntimeDecisions: 0,
    totalAvoidedRuntimeDecisions: 0,
    semanticMismatches: 0,
    heapDeltaBytes: 0,
  };
}

async function executeVariantSample(
  accumulator: BraidFastPathVariantAccumulator,
  compiled: GnosisTypeScriptBridgeResult,
  inputValue: number,
  measured: boolean
): Promise<void> {
  const heapBefore = process.memoryUsage().heapUsed;
  const start = performance.now();
  const result = await executeTypeScriptWithGnosis({
    compiled,
    input: { value: inputValue },
    bindings: BENCHMARK_BINDINGS,
    certifiedBraidFastPath: accumulator.decision,
  });
  const elapsedMs = performance.now() - start;
  const heapAfter = process.memoryUsage().heapUsed;
  const braidState = result.braidFastPath;

  if (!measured) {
    return;
  }

  accumulator.timingsMs.push(elapsedMs);
  if (result.payload !== expectedPayload(inputValue)) {
    accumulator.semanticMismatches += 1;
  }

  accumulator.status = braidState?.status ?? accumulator.status;
  accumulator.reason = braidState?.reason ?? accumulator.reason;
  accumulator.totalRuntimeDecisions += braidState?.runtimeDecisionCount ?? 0;
  accumulator.totalAvoidedRuntimeDecisions +=
    braidState?.avoidedRuntimeDecisionCount ?? 0;
  accumulator.heapDeltaBytes += heapAfter - heapBefore;
}

function finalizeVariantReport(
  accumulator: BraidFastPathVariantAccumulator,
  iterations: number
): BraidFastPathVariantReport {
  return {
    name: accumulator.name,
    status: accumulator.status,
    reason: accumulator.reason,
    iterations,
    timingsMs: accumulator.timingsMs,
    meanMs: mean(accumulator.timingsMs),
    stdevMs: stdev(accumulator.timingsMs),
    ci95: bootstrapMeanConfidenceInterval(accumulator.timingsMs),
    totalRuntimeDecisions: accumulator.totalRuntimeDecisions,
    totalAvoidedRuntimeDecisions: accumulator.totalAvoidedRuntimeDecisions,
    semanticMismatches: accumulator.semanticMismatches,
    heapDeltaBytes: accumulator.heapDeltaBytes,
  };
}

/**
 * Handles the gnosis run Braid Fast Path Benchmark workflow.
 */
export async function runBraidFastPathBenchmark(
  options: BraidFastPathBenchmarkOptions = {}
): Promise<BraidFastPathBenchmarkReport> {
  const normalized = normalizeBenchmarkOptions(options);
  const compiled = compileBenchmarkFixture();
  const runtimeModuleSource = compiled.runtimeModuleSource;
  const selectedDecision = createSelectedDecision();
  const fallbackDecision = selectCertifiedBraidFastPath(null);
  const braidFoldCertified = selectedDecision.path !== null;
  const decompositionCertificate =
    selectedDecision.path === null
      ? null
      : createDecompositionTripleCertificate(
          {
            cycle: selectedDecision.path.certificate.cycle,
            coordinates: [7, 17, 22],
          },
          selectedDecision.path.certificate
        );
  const braidPacketCertificate =
    selectedDecision.path === null
      ? null
      : createBraidPacketCertificate(selectedDecision.path.certificate.cycle);
  const extendedAeonCertificate =
    braidPacketCertificate === null
      ? null
      : createExtendedAeonTransitionCertificate(braidPacketCertificate);
  const interferePortCertificate =
    extendedAeonCertificate === null
      ? null
      : createGnosisUringInterferePortCertificate(extendedAeonCertificate);
  const interferenceStrideCertificate =
    braidPacketCertificate === null
      ? null
      : createInterferenceStrideCertificate(braidPacketCertificate);
  const pressureWeightedInterferenceCertificate =
    interferenceStrideCertificate === null
      ? null
      : createPressureWeightedInterferenceStrideCertificate(
          interferenceStrideCertificate
        );
  const selectedAccumulator = createVariantAccumulator(
    'certified-braid-fast-path',
    selectedDecision
  );
  const fallbackAccumulator = createVariantAccumulator(
    'linear-fallback',
    fallbackDecision
  );

  for (
    let index = 0;
    index < normalized.warmup + normalized.iterations;
    index += 1
  ) {
    const measured = index >= normalized.warmup;
    const inputValue = 7 + (index % 29);
    const selectedFirst = index % 2 === 0;
    const first = selectedFirst ? selectedAccumulator : fallbackAccumulator;
    const second = selectedFirst ? fallbackAccumulator : selectedAccumulator;

    await executeVariantSample(first, compiled, inputValue, measured);
    await executeVariantSample(second, compiled, inputValue, measured);
  }

  const selected = finalizeVariantReport(
    selectedAccumulator,
    normalized.iterations
  );
  const fallback = finalizeVariantReport(
    fallbackAccumulator,
    normalized.iterations
  );

  return {
    label: 'gnosis-braid-fast-path-benchmark-v1',
    iterations: normalized.iterations,
    warmup: normalized.warmup,
    scheduleKinds: compiled.schedule.map((wave) => wave.kind),
    runtimeBindingNames: compiled.runtimeBindingNames,
    runtimeModuleMetadata: {
      emitted: runtimeModuleSource !== null,
      braidFoldDirective:
        runtimeModuleSource?.includes('@gnosis-braid-fold') ?? false,
      decompositionTripleDirective:
        runtimeModuleSource?.includes('@gnosis-decomposition-triple') ?? false,
      braidPacketDirective:
        runtimeModuleSource?.includes('@gnosis-braid-packet') ?? false,
      extendedAeonDirective:
        runtimeModuleSource?.includes('@gnosis-extended-aeon-transition') ??
        false,
      interferePortDirective:
        runtimeModuleSource?.includes('@gnosis-interfere-port') ?? false,
      pressureWeightedInterferenceDirective:
        runtimeModuleSource?.includes(
          '@gnosis-pressure-weighted-interference'
        ) ?? false,
      braidFoldCertified,
      decompositionTripleCertified: decompositionCertificate !== null,
      braidPacketCertified: braidPacketCertificate !== null,
      extendedAeonCertified: extendedAeonCertificate !== null,
      interferePortCertified: interferePortCertificate !== null,
      pressureWeightedInterferenceCertified:
        pressureWeightedInterferenceCertificate !== null,
    },
    selected,
    fallback,
    comparison: {
      meanSpeedupRatio:
        selected.meanMs === 0 ? 0 : fallback.meanMs / selected.meanMs,
      meanDeltaMs: fallback.meanMs - selected.meanMs,
      runtimeDecisionDelta:
        fallback.totalRuntimeDecisions - selected.totalRuntimeDecisions,
      avoidedRuntimeDecisionDelta:
        selected.totalAvoidedRuntimeDecisions -
        fallback.totalAvoidedRuntimeDecisions,
      semanticMismatches:
        selected.semanticMismatches + fallback.semanticMismatches,
    },
  };
}

/**
 * Runs the gnosis command-line workflow.
 */
export async function main(): Promise<BraidFastPathBenchmarkReport> {
  const argv = process.argv.slice(2);
  const report = await runBraidFastPathBenchmark({
    iterations: parseNumericFlag(argv, 'iterations', 100),
    warmup: parseNumericFlag(argv, 'warmup', 10),
  });
  console.log(JSON.stringify(report, null, 2));
  return report;
}

const invokedPath = process.argv[1];
if (invokedPath && import.meta.url === pathToFileURL(invokedPath).href) {
  main().catch((error: unknown) => {
    console.error(error);
    process.exitCode = 1;
  });
}
