// @a0n/gnosis/src/benchmarks/compiler-phase-benchmark.ts
import { readFile, readdir } from 'node:fs/promises';
import { join, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import { performance } from 'node:perf_hooks';

import { parseGgProgram } from '@a0n/aeon-logic';
import { lowerUfcsSource } from '../ufcs';
import { BettyCompiler } from '../betty/compiler';
import { runBettiPipeline } from '../betti/bootstrap';
import {
  parseNodeDeclarations,
  parseEdgeDeclarations,
  parseProperties,
  stripComments,
} from '../betty/parse-utils';
import type { GraphAST, ASTNode, ASTEdge } from '../betty/compiler';
import { analyzeTopologyStability } from '../betty/stability';
import { createDefaultOptimizer } from '../betty/optimizer';
import {
  extractFiberPartition,
  validatePartition,
  synthesizeCoarsening,
} from '../betty/coarsen';
import { analyzeSemanticCompatibility } from '../betty/semantic-compatibility';
import { generateHopeCertificate } from '../betty/semantic-hope';
import { generateLeanFromGnosisAst } from '../betty/lean';
import {
  computeVoidDimensions,
  computeLandauerHeat,
  computeDeficitAnalysis,
} from '../betty/verify';

import {
  mean,
  stdev,
  bootstrapMeanConfidenceInterval,
  type ConfidenceInterval,
} from './statistics';

// ═══════════════════════════════════════════════════════════════════════════════
// Types
// ═══════════════════════════════════════════════════════════════════════════════

export type CompilerVariant =
  | 'aeon-logic' // lightweight regex parser (parseGgProgram)
  | 'betty' // professional compiler with all passes (BettyCompiler.parse)
  | 'betti' // self-hosted compiler (runBettiPipeline)
  | 'franky' // polyglot ditto compiler: FORK(6 detectors) → RACE → compile
  | 'beckett'; // transport compiler: FORK(chunks) → FORK(codecs) → RACE → FOLD

export type CompilerPhase =
  | 'ufcs-lowering'
  | 'aeon-logic-parse'
  | 'betty-parse-full'
  | 'betti-pipeline'
  | 'stability'
  | 'optimizer'
  | 'coarsening'
  | 'semantic-compatibility'
  | 'hope-certificate'
  | 'void-dimensions'
  | 'landauer-heat'
  | 'deficit-analysis'
  | 'lean-codegen';

export interface PhaseTimingResult {
  readonly phase: CompilerPhase;
  readonly timingsMs: readonly number[];
  readonly meanMs: number;
  readonly stdevMs: number;
  readonly ci95: ConfidenceInterval;
  readonly medianMs: number;
  readonly minMs: number;
  readonly maxMs: number;
}

export interface TopologySizeClass {
  readonly label: string;
  readonly files: readonly string[];
  readonly meanLineCount: number;
  readonly meanNodeCount: number;
  readonly meanEdgeCount: number;
}

export interface TopologyPhaseReport {
  readonly topologyPath: string;
  readonly lineCount: number;
  readonly nodeCount: number;
  readonly edgeCount: number;
  readonly beta1: number;
  readonly phases: readonly PhaseTimingResult[];
}

export interface SizeClassReport {
  readonly sizeClass: TopologySizeClass;
  readonly topologyReports: readonly TopologyPhaseReport[];
  readonly aggregatePhases: readonly PhaseTimingResult[];
}

export interface CompilerPhaseBenchmarkConfig {
  readonly label: 'gnosis-compiler-phase-benchmark-v1';
  readonly iterations: number;
  readonly warmupIterations: number;
}

export interface CompilerShootoutEntry {
  readonly compiler: CompilerVariant;
  readonly timingsMs: readonly number[];
  readonly meanMs: number;
  readonly stdevMs: number;
  readonly ci95: ConfidenceInterval;
  readonly medianMs: number;
  readonly nodeCount: number;
  readonly edgeCount: number;
  readonly beta1: number;
}

export interface NamedTopologyShootout {
  readonly topologyName: string;
  readonly lineCount: number;
  readonly compilers: readonly CompilerShootoutEntry[];
  readonly winner: CompilerVariant;
  readonly winnerSpeedup: number;
}

export interface CompilerPhaseBenchmarkReport {
  readonly label: CompilerPhaseBenchmarkConfig['label'];
  readonly config: CompilerPhaseBenchmarkConfig;
  readonly sizeClasses: readonly SizeClassReport[];
  readonly compilerShootout: readonly NamedTopologyShootout[];
  readonly headToHead: {
    readonly aeonLogicVsBetty: readonly {
      readonly sizeClass: string;
      readonly aeonLogicMeanMs: number;
      readonly bettyFullMeanMs: number;
      readonly overhead: number;
    }[];
    readonly compilerRanking: readonly {
      readonly compiler: CompilerVariant;
      readonly meanMs: number;
      readonly fractionOfSlowest: number;
    }[];
    readonly phaseCostRanking: readonly {
      readonly phase: CompilerPhase;
      readonly meanMs: number;
      readonly fractionOfTotal: number;
    }[];
  };
}

// ═══════════════════════════════════════════════════════════════════════════════
// Default configuration
// ═══════════════════════════════════════════════════════════════════════════════

export const DEFAULT_CONFIG: CompilerPhaseBenchmarkConfig = {
  label: 'gnosis-compiler-phase-benchmark-v1',
  iterations: 20,
  warmupIterations: 3,
};

export const REDUCED_CONFIG: CompilerPhaseBenchmarkConfig = {
  label: 'gnosis-compiler-phase-benchmark-v1',
  iterations: 5,
  warmupIterations: 1,
};

// ═══════════════════════════════════════════════════════════════════════════════
// Helpers
// ═══════════════════════════════════════════════════════════════════════════════

function median(values: readonly number[]): number {
  if (values.length === 0) return 0;
  const sorted = [...values].sort((a, b) => a - b);
  const mid = Math.floor(sorted.length / 2);
  return sorted.length % 2 === 0
    ? ((sorted[mid - 1] ?? 0) + (sorted[mid] ?? 0)) / 2
    : sorted[mid] ?? 0;
}

function timePhase(
  fn: () => void,
  iterations: number,
  warmup: number
): number[] {
  // Warmup
  for (let i = 0; i < warmup; i++) {
    fn();
  }

  const timings: number[] = [];
  for (let i = 0; i < iterations; i++) {
    const start = performance.now();
    fn();
    const end = performance.now();
    timings.push(end - start);
  }
  return timings;
}

function buildPhaseResult(
  phase: CompilerPhase,
  timings: number[]
): PhaseTimingResult {
  return {
    phase,
    timingsMs: timings,
    meanMs: mean(timings),
    stdevMs: stdev(timings),
    ci95: bootstrapMeanConfidenceInterval(timings),
    medianMs: median(timings),
    minMs: Math.min(...timings),
    maxMs: Math.max(...timings),
  };
}

// ═══════════════════════════════════════════════════════════════════════════════
// Topology discovery
// ═══════════════════════════════════════════════════════════════════════════════

const GNOSIS_ROOT = resolve(fileURLToPath(import.meta.url), '..', '..', '..');

/**
 * Handles the gnosis discover Topologies workflow.
 */
export async function discoverTopologies(): Promise<{
  small: string[];
  medium: string[];
  large: string[];
}> {
  const rootFiles = await readdir(GNOSIS_ROOT);
  const ggFiles = rootFiles
    .filter((f) => f.endsWith('.gg') && !f.endsWith('.test.gg'))
    .map((f) => join(GNOSIS_ROOT, f));

  // Read line counts
  const fileSizes: { path: string; lines: number }[] = [];
  for (const filePath of ggFiles) {
    const content = await readFile(filePath, 'utf-8');
    fileSizes.push({ path: filePath, lines: content.split('\n').length });
  }
  fileSizes.sort((a, b) => a.lines - b.lines);

  // Partition into thirds
  const third = Math.floor(fileSizes.length / 3);
  return {
    small: fileSizes.slice(0, third).map((f) => f.path),
    medium: fileSizes.slice(third, third * 2).map((f) => f.path),
    large: fileSizes.slice(third * 2).map((f) => f.path),
  };
}

/**
 * Handles the gnosis select Representative workflow.
 */
export function selectRepresentative(files: string[], count: number): string[] {
  if (files.length <= count) return files;
  const step = Math.floor(files.length / count);
  const result: string[] = [];
  for (let i = 0; i < count; i++) {
    result.push(files[Math.min(i * step, files.length - 1)] as string);
  }
  return result;
}

// ═══════════════════════════════════════════════════════════════════════════════
// Topology-as-compiler runners
//
// Each .gg topology defines a compilation strategy. These functions execute
// that strategy on .gg input, using the shared parse-utils for real work.
// The difference is the fork/race/fold *shape* of the pipeline.
// ═══════════════════════════════════════════════════════════════════════════════

/**
 * Franky pipeline: polyglot racing compiler.
 *
 * Shape: source → FORK(N detectors) → RACE(first match) → compile → emit
 *
 * Franky's insight: fork into multiple parallel detection/parsing paths,
 * race them, and the first to finish wins. Applied to .gg compilation:
 * fork into N parallel parsers with different strategies, race to first
 * valid AST.
 */
export function runFrankyPipeline(source: string): GraphAST {
  const stripped = stripComments(source);
  const lines = stripped.split('\n').filter((l) => l.trim());

  // FORK: six parallel "detector" paths parse the source simultaneously
  // Each path uses a different entry strategy (mirroring Franky's 6 framework detectors)
  const paths: GraphAST[] = [];

  // Path 1: Node-first parse (like Express detection -- find app names first)
  {
    const nodes = new Map<string, ASTNode>();
    const edges: ASTEdge[] = [];
    for (const line of lines) {
      for (const pn of parseNodeDeclarations(line.trim())) {
        if (!pn.id) continue;
        nodes.set(pn.id, {
          id: pn.id,
          labels: pn.label ? [pn.label] : [],
          properties: parseProperties(pn.propertiesRaw),
        });
      }
    }
    for (const line of lines) {
      for (const pe of parseEdgeDeclarations(line.trim())) {
        const sourceIds = pe.sourceRaw
          .split('|')
          .map((s) => s.split(':')[0].trim());
        const targetIds = pe.targetRaw
          .split('|')
          .map((s) => s.split(':')[0].trim());
        edges.push({
          sourceIds,
          targetIds,
          type: pe.edgeType,
          properties: parseProperties(pe.propertiesRaw),
        });
        for (const id of [...sourceIds, ...targetIds]) {
          if (!nodes.has(id)) nodes.set(id, { id, labels: [], properties: {} });
        }
      }
    }
    paths.push({ nodes, edges });
  }

  // Path 2: Edge-first parse (like Gin detection -- find routes first)
  {
    const nodes = new Map<string, ASTNode>();
    const edges: ASTEdge[] = [];
    for (const line of lines) {
      for (const pe of parseEdgeDeclarations(line.trim())) {
        const sourceIds = pe.sourceRaw
          .split('|')
          .map((s) => s.split(':')[0].trim());
        const targetIds = pe.targetRaw
          .split('|')
          .map((s) => s.split(':')[0].trim());
        edges.push({
          sourceIds,
          targetIds,
          type: pe.edgeType,
          properties: parseProperties(pe.propertiesRaw),
        });
        for (const id of [...sourceIds, ...targetIds]) {
          if (!nodes.has(id)) nodes.set(id, { id, labels: [], properties: {} });
        }
      }
    }
    for (const line of lines) {
      for (const pn of parseNodeDeclarations(line.trim())) {
        if (!pn.id) continue;
        const existing = nodes.get(pn.id);
        if (existing) {
          if (pn.label && existing.labels.length === 0)
            existing.labels = [pn.label];
          existing.properties = {
            ...existing.properties,
            ...parseProperties(pn.propertiesRaw),
          };
        } else {
          nodes.set(pn.id, {
            id: pn.id,
            labels: pn.label ? [pn.label] : [],
            properties: parseProperties(pn.propertiesRaw),
          });
        }
      }
    }
    paths.push({ nodes, edges });
  }

  // Path 3-6: Chunked parallel parse (like Franky's 4 additional framework detectors)
  // Split input into chunks and parse each independently, then merge
  const chunkCount = 4;
  const chunkSize = Math.ceil(lines.length / chunkCount);
  const chunkAsts: GraphAST[] = [];
  for (let i = 0; i < chunkCount; i++) {
    const chunk = lines.slice(i * chunkSize, (i + 1) * chunkSize);
    const nodes = new Map<string, ASTNode>();
    const edges: ASTEdge[] = [];
    for (const line of chunk) {
      const trimmed = line.trim();
      for (const pn of parseNodeDeclarations(trimmed)) {
        if (!pn.id) continue;
        nodes.set(pn.id, {
          id: pn.id,
          labels: pn.label ? [pn.label] : [],
          properties: parseProperties(pn.propertiesRaw),
        });
      }
      for (const pe of parseEdgeDeclarations(trimmed)) {
        const sourceIds = pe.sourceRaw
          .split('|')
          .map((s) => s.split(':')[0].trim());
        const targetIds = pe.targetRaw
          .split('|')
          .map((s) => s.split(':')[0].trim());
        edges.push({
          sourceIds,
          targetIds,
          type: pe.edgeType,
          properties: parseProperties(pe.propertiesRaw),
        });
        for (const id of [...sourceIds, ...targetIds]) {
          if (!nodes.has(id)) nodes.set(id, { id, labels: [], properties: {} });
        }
      }
    }
    chunkAsts.push({ nodes, edges });
  }

  // FOLD: merge all chunk ASTs into one
  const merged = new Map<string, ASTNode>();
  const mergedEdges: ASTEdge[] = [];
  for (const chunk of chunkAsts) {
    for (const [id, node] of chunk.nodes) {
      if (!merged.has(id)) merged.set(id, { ...node });
      else {
        const existing = merged.get(id)!;
        if (node.labels.length > 0 && existing.labels.length === 0)
          existing.labels = node.labels;
        existing.properties = { ...existing.properties, ...node.properties };
      }
    }
    mergedEdges.push(...chunk.edges);
  }
  paths.push({ nodes: merged, edges: mergedEdges });

  // RACE: first valid AST wins (the one with most edges -- most complete parse)
  paths.sort((a, b) => b.edges.length - a.edges.length);
  return paths[0] ?? { nodes: new Map(), edges: [] };
}

/**
 * Beckett pipeline: transport/streaming compiler.
 *
 * Shape: source → serialize → FORK(chunks) → FORK(codecs) → RACE → FOLD(reassemble)
 *
 * Beckett's insight: chunk the input, process each chunk through multiple
 * parallel strategies, race per-chunk, then fold back together. Applied to
 * .gg compilation: chunk the source, parse each chunk with N strategies,
 * race to best per-chunk parse, reassemble.
 */
export function runBeckettPipeline(source: string): GraphAST {
  const stripped = stripComments(source);
  const lines = stripped.split('\n').filter((l) => l.trim());

  // STAGE 1: Serialize -- flatten the source into processable units
  // (like Beckett's TopologySerializer → AeonFlowFrames)

  // STAGE 2: Chunk -- split into independent chunks
  // (like Beckett's LaminarChunker with chunk_size: 65536)
  const chunkSize = Math.max(4, Math.ceil(lines.length / 4));
  const chunks: string[][] = [];
  for (let i = 0; i < lines.length; i += chunkSize) {
    chunks.push(lines.slice(i, i + chunkSize));
  }

  // STAGE 3: Codec race -- per-chunk, fork to multiple parse strategies, race
  // (like Beckett's identity | gzip | brotli | deflate race)
  const chunkResults: { nodes: Map<string, ASTNode>; edges: ASTEdge[] }[] = [];

  for (const chunk of chunks) {
    // Fork: four "codecs" (parse strategies) per chunk
    const strategies: {
      nodes: Map<string, ASTNode>;
      edges: ASTEdge[];
      score: number;
    }[] = [];

    // Strategy 1: "identity" -- raw parse, no optimization
    {
      const nodes = new Map<string, ASTNode>();
      const edges: ASTEdge[] = [];
      for (const line of chunk) {
        const trimmed = line.trim();
        for (const pn of parseNodeDeclarations(trimmed)) {
          if (pn.id)
            nodes.set(pn.id, {
              id: pn.id,
              labels: pn.label ? [pn.label] : [],
              properties: parseProperties(pn.propertiesRaw),
            });
        }
        for (const pe of parseEdgeDeclarations(trimmed)) {
          const sourceIds = pe.sourceRaw
            .split('|')
            .map((s) => s.split(':')[0].trim());
          const targetIds = pe.targetRaw
            .split('|')
            .map((s) => s.split(':')[0].trim());
          edges.push({
            sourceIds,
            targetIds,
            type: pe.edgeType,
            properties: parseProperties(pe.propertiesRaw),
          });
          for (const id of [...sourceIds, ...targetIds]) {
            if (!nodes.has(id))
              nodes.set(id, { id, labels: [], properties: {} });
          }
        }
      }
      strategies.push({ nodes, edges, score: edges.length + nodes.size });
    }

    // Strategy 2: "gzip" -- edges first then nodes (reordered parse)
    {
      const nodes = new Map<string, ASTNode>();
      const edges: ASTEdge[] = [];
      for (const line of chunk) {
        for (const pe of parseEdgeDeclarations(line.trim())) {
          const sourceIds = pe.sourceRaw
            .split('|')
            .map((s) => s.split(':')[0].trim());
          const targetIds = pe.targetRaw
            .split('|')
            .map((s) => s.split(':')[0].trim());
          edges.push({
            sourceIds,
            targetIds,
            type: pe.edgeType,
            properties: parseProperties(pe.propertiesRaw),
          });
          for (const id of [...sourceIds, ...targetIds]) {
            if (!nodes.has(id))
              nodes.set(id, { id, labels: [], properties: {} });
          }
        }
      }
      for (const line of chunk) {
        for (const pn of parseNodeDeclarations(line.trim())) {
          if (pn.id && !nodes.has(pn.id))
            nodes.set(pn.id, {
              id: pn.id,
              labels: pn.label ? [pn.label] : [],
              properties: parseProperties(pn.propertiesRaw),
            });
        }
      }
      strategies.push({ nodes, edges, score: edges.length + nodes.size });
    }

    // Strategy 3: "brotli" -- full Betti-style interleaved parse
    {
      const nodes = new Map<string, ASTNode>();
      const edges: ASTEdge[] = [];
      for (const line of chunk) {
        const trimmed = line.trim();
        const parsedNodes = parseNodeDeclarations(trimmed);
        const parsedEdges = parseEdgeDeclarations(trimmed);
        for (const pn of parsedNodes) {
          if (pn.id) {
            const props = parseProperties(pn.propertiesRaw);
            if (!nodes.has(pn.id))
              nodes.set(pn.id, {
                id: pn.id,
                labels: pn.label ? [pn.label] : [],
                properties: props,
              });
            else {
              const e = nodes.get(pn.id)!;
              if (pn.label && !e.labels.length) e.labels = [pn.label];
              e.properties = { ...e.properties, ...props };
            }
          }
        }
        for (const pe of parsedEdges) {
          const sourceIds = pe.sourceRaw
            .split('|')
            .map((s) => s.split(':')[0].trim());
          const targetIds = pe.targetRaw
            .split('|')
            .map((s) => s.split(':')[0].trim());
          edges.push({
            sourceIds,
            targetIds,
            type: pe.edgeType,
            properties: parseProperties(pe.propertiesRaw),
          });
          for (const id of [...sourceIds, ...targetIds]) {
            if (!nodes.has(id))
              nodes.set(id, { id, labels: [], properties: {} });
          }
        }
      }
      strategies.push({ nodes, edges, score: edges.length + nodes.size });
    }

    // Strategy 4: "deflate" -- reverse-order parse
    {
      const nodes = new Map<string, ASTNode>();
      const edges: ASTEdge[] = [];
      for (let i = chunk.length - 1; i >= 0; i--) {
        const trimmed = chunk[i].trim();
        for (const pn of parseNodeDeclarations(trimmed)) {
          if (pn.id)
            nodes.set(pn.id, {
              id: pn.id,
              labels: pn.label ? [pn.label] : [],
              properties: parseProperties(pn.propertiesRaw),
            });
        }
        for (const pe of parseEdgeDeclarations(trimmed)) {
          const sourceIds = pe.sourceRaw
            .split('|')
            .map((s) => s.split(':')[0].trim());
          const targetIds = pe.targetRaw
            .split('|')
            .map((s) => s.split(':')[0].trim());
          edges.push({
            sourceIds,
            targetIds,
            type: pe.edgeType,
            properties: parseProperties(pe.propertiesRaw),
          });
          for (const id of [...sourceIds, ...targetIds]) {
            if (!nodes.has(id))
              nodes.set(id, { id, labels: [], properties: {} });
          }
        }
      }
      strategies.push({ nodes, edges, score: edges.length + nodes.size });
    }

    // RACE: best score wins (most complete parse)
    strategies.sort((a, b) => b.score - a.score);
    chunkResults.push(strategies[0]!);
  }

  // STAGE 5: FOLD(reassemble) -- merge all chunk results in order
  const finalNodes = new Map<string, ASTNode>();
  const finalEdges: ASTEdge[] = [];
  for (const chunk of chunkResults) {
    for (const [id, node] of chunk.nodes) {
      if (!finalNodes.has(id)) finalNodes.set(id, { ...node });
      else {
        const existing = finalNodes.get(id)!;
        if (node.labels.length > 0 && existing.labels.length === 0)
          existing.labels = node.labels;
        existing.properties = { ...existing.properties, ...node.properties };
      }
    }
    finalEdges.push(...chunk.edges);
  }

  return { nodes: finalNodes, edges: finalEdges };
}

// ═══════════════════════════════════════════════════════════════════════════════
// Phase benchmarking
// ═══════════════════════════════════════════════════════════════════════════════

/**
 * Handles the gnosis benchmark Topology workflow.
 */
export function benchmarkTopology(
  source: string,
  config: CompilerPhaseBenchmarkConfig
): TopologyPhaseReport {
  const lines = source.split('\n');
  const { iterations, warmupIterations } = config;

  // Phase 1: UFCS lowering
  let lowered = '';
  const ufcsTimings = timePhase(
    () => {
      lowered = lowerUfcsSource(source);
    },
    iterations,
    warmupIterations
  );

  // Phase 2: aeon-logic parseGgProgram (the lightweight parser)
  const aeonLogicTimings = timePhase(
    () => {
      parseGgProgram(lowered);
    },
    iterations,
    warmupIterations
  );

  // Phase 3: Betty full parse (the professional compiler)
  const betty = new BettyCompiler();
  let bettyResult = betty.parse(source); // initial parse for downstream phases
  const bettyFullTimings = timePhase(
    () => {
      const b = new BettyCompiler();
      bettyResult = b.parse(source);
    },
    iterations,
    warmupIterations
  );

  // Phase 4: Betti self-hosted pipeline
  const bettiTimings = timePhase(
    () => {
      runBettiPipeline(source);
    },
    iterations,
    warmupIterations
  );

  // Phase 5: Stability analysis (isolated)
  const ast = bettyResult.ast;
  const b1 = bettyResult.b1;
  let stabilityReport = bettyResult.stability;
  const stabilityTimings = ast
    ? timePhase(
        () => {
          stabilityReport = analyzeTopologyStability(ast, b1);
        },
        iterations,
        warmupIterations
      )
    : [];

  // Phase 6: Optimizer passes (isolated)
  const optimizerTimings =
    ast && stabilityReport
      ? timePhase(
          () => {
            const opt = createDefaultOptimizer();
            opt.apply(ast, stabilityReport!);
          },
          iterations,
          warmupIterations
        )
      : [];

  // Phase 7: Coarsening synthesis (isolated)
  const coarseningTimings =
    ast && stabilityReport
      ? timePhase(
          () => {
            const partition = extractFiberPartition(ast);
            if (partition) {
              const valDiags = validatePartition(partition, stabilityReport!);
              if (!valDiags.some((d) => d.severity === 'error')) {
                synthesizeCoarsening(ast, stabilityReport!, partition);
              }
            }
          },
          iterations,
          warmupIterations
        )
      : [];

  // Phase 8: Semantic compatibility (isolated)
  let semanticResult = bettyResult.semanticCompatibility;
  const semanticTimings = ast
    ? timePhase(
        () => {
          semanticResult = analyzeSemanticCompatibility(ast);
        },
        iterations,
        warmupIterations
      )
    : [];

  // Phase 9: Hope certificate (isolated)
  const hopeTimings =
    ast && semanticResult
      ? timePhase(
          () => {
            generateHopeCertificate(ast, semanticResult!);
          },
          iterations,
          warmupIterations
        )
      : [];

  // Phase 10: Void dimensions (isolated)
  const voidDimTimings = ast
    ? timePhase(
        () => {
          computeVoidDimensions(ast);
        },
        iterations,
        warmupIterations
      )
    : [];

  // Phase 11: Landauer heat (isolated)
  const landauerTimings = ast
    ? timePhase(
        () => {
          computeLandauerHeat(ast);
        },
        iterations,
        warmupIterations
      )
    : [];

  // Phase 12: Deficit analysis (isolated)
  const deficitTimings = ast
    ? timePhase(
        () => {
          computeDeficitAnalysis(ast);
        },
        iterations,
        warmupIterations
      )
    : [];

  // Phase 13: Lean codegen (isolated)
  const leanTimings =
    ast && stabilityReport
      ? timePhase(
          () => {
            generateLeanFromGnosisAst(ast, stabilityReport!, {});
          },
          iterations,
          warmupIterations
        )
      : [];

  const phases: PhaseTimingResult[] = [
    buildPhaseResult('ufcs-lowering', ufcsTimings),
    buildPhaseResult('aeon-logic-parse', aeonLogicTimings),
    buildPhaseResult('betty-parse-full', bettyFullTimings),
    buildPhaseResult('betti-pipeline', bettiTimings),
    buildPhaseResult('stability', stabilityTimings),
    buildPhaseResult('optimizer', optimizerTimings),
    buildPhaseResult('coarsening', coarseningTimings),
    buildPhaseResult('semantic-compatibility', semanticTimings),
    buildPhaseResult('hope-certificate', hopeTimings),
    buildPhaseResult('void-dimensions', voidDimTimings),
    buildPhaseResult('landauer-heat', landauerTimings),
    buildPhaseResult('deficit-analysis', deficitTimings),
    buildPhaseResult('lean-codegen', leanTimings),
  ];

  return {
    topologyPath: '',
    lineCount: lines.length,
    nodeCount: ast?.nodes.size ?? 0,
    edgeCount: ast?.edges.length ?? 0,
    beta1: b1,
    phases,
  };
}

// ═══════════════════════════════════════════════════════════════════════════════
// Size class aggregation
// ═══════════════════════════════════════════════════════════════════════════════

function aggregatePhases(
  reports: readonly TopologyPhaseReport[]
): PhaseTimingResult[] {
  if (reports.length === 0) return [];

  const phaseNames = reports[0]!.phases.map((p) => p.phase);
  return phaseNames.map((phase) => {
    const allTimings = reports.flatMap(
      (r) => r.phases.find((p) => p.phase === phase)?.timingsMs ?? []
    );
    return buildPhaseResult(phase, [...allTimings]);
  });
}

/**
 * Handles the gnosis benchmark Size Class workflow.
 */
export async function benchmarkSizeClass(
  label: string,
  files: string[],
  samplesPerClass: number,
  config: CompilerPhaseBenchmarkConfig
): Promise<SizeClassReport> {
  const selected = selectRepresentative(files, samplesPerClass);
  const topologyReports: TopologyPhaseReport[] = [];

  let totalLines = 0;
  let totalNodes = 0;
  let totalEdges = 0;

  for (const filePath of selected) {
    const source = await readFile(filePath, 'utf-8');
    const report = benchmarkTopology(source, config);
    topologyReports.push({
      ...report,
      topologyPath: filePath,
    });
    totalLines += report.lineCount;
    totalNodes += report.nodeCount;
    totalEdges += report.edgeCount;
  }

  const count = selected.length || 1;
  return {
    sizeClass: {
      label,
      files: selected,
      meanLineCount: totalLines / count,
      meanNodeCount: totalNodes / count,
      meanEdgeCount: totalEdges / count,
    },
    topologyReports,
    aggregatePhases: aggregatePhases(topologyReports),
  };
}

// ═══════════════════════════════════════════════════════════════════════════════
// Head-to-head comparison
// ═══════════════════════════════════════════════════════════════════════════════

function buildHeadToHead(
  sizeClasses: readonly SizeClassReport[]
): Pick<
  CompilerPhaseBenchmarkReport['headToHead'],
  'aeonLogicVsBetty' | 'phaseCostRanking'
> {
  const aeonLogicVsBetty = sizeClasses.map((sc) => {
    const aeonLogic = sc.aggregatePhases.find(
      (p) => p.phase === 'aeon-logic-parse'
    );
    const bettyFull = sc.aggregatePhases.find(
      (p) => p.phase === 'betty-parse-full'
    );
    const aeonMs = aeonLogic?.meanMs ?? 0;
    const bettyMs = bettyFull?.meanMs ?? 0;
    return {
      sizeClass: sc.sizeClass.label,
      aeonLogicMeanMs: aeonMs,
      bettyFullMeanMs: bettyMs,
      overhead: aeonMs > 0 ? bettyMs / aeonMs : 0,
    };
  });

  // Phase cost ranking (averaged across all size classes)
  const allPhases = new Map<CompilerPhase, number[]>();
  for (const sc of sizeClasses) {
    for (const phase of sc.aggregatePhases) {
      const existing = allPhases.get(phase.phase) ?? [];
      existing.push(phase.meanMs);
      allPhases.set(phase.phase, existing);
    }
  }

  const totalMean = Array.from(allPhases.values())
    .map((vals) => mean(vals))
    .reduce((sum, v) => sum + v, 0);

  const phaseCostRanking = Array.from(allPhases.entries())
    .map(([phase, vals]) => ({
      phase,
      meanMs: mean(vals),
      fractionOfTotal: totalMean > 0 ? mean(vals) / totalMean : 0,
    }))
    .sort((a, b) => b.meanMs - a.meanMs);

  return { aeonLogicVsBetty, phaseCostRanking };
}

// ═══════════════════════════════════════════════════════════════════════════════
// Main report generator
// ═══════════════════════════════════════════════════════════════════════════════

/**
 * Handles the gnosis run Compiler Phase Benchmark workflow.
 */
export async function runCompilerPhaseBenchmark(
  config: CompilerPhaseBenchmarkConfig = DEFAULT_CONFIG,
  samplesPerClass: number = 3
): Promise<CompilerPhaseBenchmarkReport> {
  const topology = await discoverTopologies();

  const [smallReport, mediumReport, largeReport, compilerShootout] =
    await Promise.all([
      benchmarkSizeClass('small', topology.small, samplesPerClass, config),
      benchmarkSizeClass('medium', topology.medium, samplesPerClass, config),
      benchmarkSizeClass('large', topology.large, samplesPerClass, config),
      runCompilerShootout(config),
    ]);

  const sizeClasses = [smallReport, mediumReport, largeReport];
  const headToHead = buildHeadToHead(sizeClasses);

  return {
    label: config.label,
    config,
    sizeClasses,
    compilerShootout,
    headToHead: {
      ...headToHead,
      compilerRanking: buildCompilerRanking(compilerShootout),
    },
  };
}

// ═══════════════════════════════════════════════════════════════════════════════
// Inline topologies for deterministic testing
// ═══════════════════════════════════════════════════════════════════════════════

export const INLINE_SMALL = `// small -- 3 nodes, 2 edges
(src: Source)
(src)-[:FORK]->(a: Worker | b: Worker)
(a | b)-[:FOLD]->(sink: Sink)`;

export const INLINE_MEDIUM = `// medium -- fork/race/fold pipeline with stability annotations
(input: Source { arrival_rate: '10', service_rate: '15' })
(input)-[:FORK]->(w1: Worker { arrival_rate: '5', service_rate: '8' } | w2: Worker { arrival_rate: '5', service_rate: '8' } | w3: Worker { arrival_rate: '5', service_rate: '8' })
(w1 | w2 | w3)-[:RACE { take: 'fastest' }]->(winner: Selector)
(winner)-[:FOLD { strategy: 'merge' }]->(output: Sink { arrival_rate: '10', service_rate: '20' })
(w1 | w2 | w3)-[:VENT]->(void: VoidBoundary)
(output)-[:PROCESS]->(logger: Logger)
(logger)-[:PROCESS]->(done: Done)`;

export const INLINE_LARGE = `// large -- multi-stage transformer-like topology
(input: Source { arrival_rate: '10', service_rate: '20' })
(input)-[:FORK]->(head1: AttentionHead { arrival_rate: '2.5', service_rate: '5' } | head2: AttentionHead { arrival_rate: '2.5', service_rate: '5' } | head3: AttentionHead { arrival_rate: '2.5', service_rate: '5' } | head4: AttentionHead { arrival_rate: '2.5', service_rate: '5' })
(head1 | head2 | head3 | head4)-[:FOLD { strategy: 'concat' }]->(merged: MergedAttention { arrival_rate: '10', service_rate: '15' })
(merged)-[:PROCESS { op: 'layer_norm' }]->(normed: LayerNorm)
(normed)-[:FORK]->(expert1: FFNExpert { arrival_rate: '3.3', service_rate: '6' } | expert2: FFNExpert { arrival_rate: '3.3', service_rate: '6' } | expert3: FFNExpert { arrival_rate: '3.3', service_rate: '6' })
(expert1 | expert2 | expert3)-[:RACE { take: 'top-2' }]->(gate: GateRouter)
(gate)-[:FOLD { strategy: 'weighted-sum' }]->(ffn_out: FFNOutput { arrival_rate: '10', service_rate: '18' })
(expert1 | expert2 | expert3)-[:VENT]->(ffn_void: VoidBoundary)
(ffn_out)-[:PROCESS { op: 'residual' }]->(residual: Residual)
(residual)-[:FORK]->(block2_h1: AttentionHead { arrival_rate: '2.5', service_rate: '5' } | block2_h2: AttentionHead { arrival_rate: '2.5', service_rate: '5' } | block2_h3: AttentionHead { arrival_rate: '2.5', service_rate: '5' } | block2_h4: AttentionHead { arrival_rate: '2.5', service_rate: '5' })
(block2_h1 | block2_h2 | block2_h3 | block2_h4)-[:FOLD { strategy: 'concat' }]->(block2_merged: MergedAttention { arrival_rate: '10', service_rate: '15' })
(block2_merged)-[:PROCESS { op: 'layer_norm' }]->(block2_normed: LayerNorm)
(block2_normed)-[:FORK]->(block2_e1: FFNExpert { arrival_rate: '3.3', service_rate: '6' } | block2_e2: FFNExpert { arrival_rate: '3.3', service_rate: '6' } | block2_e3: FFNExpert { arrival_rate: '3.3', service_rate: '6' })
(block2_e1 | block2_e2 | block2_e3)-[:RACE { take: 'top-2' }]->(block2_gate: GateRouter)
(block2_gate)-[:FOLD { strategy: 'weighted-sum' }]->(block2_ffn: FFNOutput { arrival_rate: '10', service_rate: '18' })
(block2_e1 | block2_e2 | block2_e3)-[:VENT]->(block2_void: VoidBoundary)
(block2_ffn)-[:PROCESS { op: 'residual' }]->(output: Output)
(output)-[:PROCESS]->(logits: Logits)
(logits)-[:PROCESS]->(softmax: Softmax)
(softmax)-[:FOLD { strategy: 'argmax' }]->(prediction: Prediction)`;

// ═══════════════════════════════════════════════════════════════════════════════
// Named topology paths (the compiler family)
// ═══════════════════════════════════════════════════════════════════════════════

const EXAMPLES_DIR = resolve(GNOSIS_ROOT, 'examples');

export const NAMED_TOPOLOGIES: Record<string, string> = {
  'betti.gg': join(GNOSIS_ROOT, 'betti.gg'),
  'franky.gg': join(EXAMPLES_DIR, 'franky.gg'),
  'beckett.gg': join(EXAMPLES_DIR, 'beckett.gg'),
};

// ═══════════════════════════════════════════════════════════════════════════════
// Compiler shootout: Betty vs Betti vs parseGgProgram on named topologies
// ═══════════════════════════════════════════════════════════════════════════════

function shootoutCompiler(
  variant: CompilerVariant,
  source: string,
  iterations: number,
  warmup: number
): CompilerShootoutEntry {
  // Capture metrics from a single dedicated run (outside timing loop)
  let nodeCount = 0;
  let edgeCount = 0;
  let beta1 = 0;

  function computeB1FromAst(ast: GraphAST): number {
    let b1 = 0;
    for (const edge of ast.edges) {
      if (edge.type === 'FORK') b1 += edge.targetIds.length - 1;
      else if (
        edge.type === 'FOLD' ||
        edge.type === 'COLLAPSE' ||
        edge.type === 'OBSERVE'
      )
        b1 = Math.max(0, b1 - (edge.sourceIds.length - 1));
      else if (edge.type === 'RACE' || edge.type === 'SLIVER')
        b1 = Math.max(
          0,
          b1 - Math.max(0, edge.sourceIds.length - edge.targetIds.length)
        );
      else if (edge.type === 'VENT') b1 = Math.max(0, b1 - 1);
    }
    return b1;
  }

  switch (variant) {
    case 'aeon-logic': {
      const prog = parseGgProgram(lowerUfcsSource(source));
      // parseGgProgram returns nodes as a Map or plain object depending on build
      nodeCount =
        prog.nodes instanceof Map
          ? prog.nodes.size
          : Object.keys(prog.nodes).length;
      edgeCount = prog.edges.length;
      beta1 = 0; // parseGgProgram does not compute beta1
      break;
    }
    case 'betty': {
      const b = new BettyCompiler();
      const result = b.parse(source);
      nodeCount = result.ast?.nodes.size ?? 0;
      edgeCount = result.ast?.edges.length ?? 0;
      beta1 = result.b1;
      break;
    }
    case 'betti': {
      const ast = runBettiPipeline(source);
      nodeCount = ast.nodes.size;
      edgeCount = ast.edges.length;
      beta1 = computeB1FromAst(ast);
      break;
    }
    case 'franky': {
      const ast = runFrankyPipeline(source);
      nodeCount = ast.nodes.size;
      edgeCount = ast.edges.length;
      beta1 = computeB1FromAst(ast);
      break;
    }
    case 'beckett': {
      const ast = runBeckettPipeline(source);
      nodeCount = ast.nodes.size;
      edgeCount = ast.edges.length;
      beta1 = computeB1FromAst(ast);
      break;
    }
  }

  // Time the hot path only
  const lowered = lowerUfcsSource(source);
  const timings = timePhase(
    () => {
      switch (variant) {
        case 'aeon-logic':
          parseGgProgram(lowered);
          break;
        case 'betty': {
          const b = new BettyCompiler();
          b.parse(source);
          break;
        }
        case 'betti':
          runBettiPipeline(source);
          break;
        case 'franky':
          runFrankyPipeline(source);
          break;
        case 'beckett':
          runBeckettPipeline(source);
          break;
      }
    },
    iterations,
    warmup
  );

  return {
    compiler: variant,
    timingsMs: timings,
    meanMs: mean(timings),
    stdevMs: stdev(timings),
    ci95: bootstrapMeanConfidenceInterval(timings),
    medianMs: median(timings),
    nodeCount,
    edgeCount,
    beta1,
  };
}

/**
 * Handles the gnosis benchmark Named Topology workflow.
 */
export function benchmarkNamedTopology(
  name: string,
  source: string,
  config: CompilerPhaseBenchmarkConfig
): NamedTopologyShootout {
  const { iterations, warmupIterations } = config;
  const lineCount = source.split('\n').length;

  const compilers: CompilerShootoutEntry[] = (
    ['aeon-logic', 'betti', 'franky', 'beckett', 'betty'] as CompilerVariant[]
  ).map((variant) =>
    shootoutCompiler(variant, source, iterations, warmupIterations)
  );

  const sorted = [...compilers].sort((a, b) => a.meanMs - b.meanMs);
  const winner = sorted[0]!;
  const slowest = sorted[sorted.length - 1]!;

  return {
    topologyName: name,
    lineCount,
    compilers,
    winner: winner.compiler,
    winnerSpeedup: winner.meanMs > 0 ? slowest.meanMs / winner.meanMs : 0,
  };
}

/**
 * Handles the gnosis run Compiler Shootout workflow.
 */
export async function runCompilerShootout(
  config: CompilerPhaseBenchmarkConfig
): Promise<NamedTopologyShootout[]> {
  const results: NamedTopologyShootout[] = [];

  for (const [name, filePath] of Object.entries(NAMED_TOPOLOGIES)) {
    try {
      const source = await readFile(filePath, 'utf-8');
      results.push(benchmarkNamedTopology(name, source, config));
    } catch {
      // File not found -- skip (submodule may not be checked out)
    }
  }

  // Also run on the inline topologies for comparison
  results.push(benchmarkNamedTopology('inline-small', INLINE_SMALL, config));
  results.push(benchmarkNamedTopology('inline-medium', INLINE_MEDIUM, config));
  results.push(benchmarkNamedTopology('inline-large', INLINE_LARGE, config));

  return results;
}

// ═══════════════════════════════════════════════════════════════════════════════
// Compiler ranking (aggregated across all shootout topologies)
// ═══════════════════════════════════════════════════════════════════════════════

function buildCompilerRanking(
  shootout: readonly NamedTopologyShootout[]
): CompilerPhaseBenchmarkReport['headToHead']['compilerRanking'] {
  const compilerTimes = new Map<CompilerVariant, number[]>();
  for (const topo of shootout) {
    for (const c of topo.compilers) {
      const existing = compilerTimes.get(c.compiler) ?? [];
      existing.push(c.meanMs);
      compilerTimes.set(c.compiler, existing);
    }
  }

  const entries = Array.from(compilerTimes.entries()).map(
    ([compiler, times]) => ({
      compiler,
      meanMs: mean(times),
    })
  );
  const slowest = Math.max(...entries.map((e) => e.meanMs), 1);

  return entries
    .map((e) => ({
      ...e,
      fractionOfSlowest: e.meanMs / slowest,
    }))
    .sort((a, b) => a.meanMs - b.meanMs);
}

// ═══════════════════════════════════════════════════════════════════════════════
// CLI entry point
// ═══════════════════════════════════════════════════════════════════════════════

if (
  typeof process !== 'undefined' &&
  process.argv[1] &&
  process.argv[1].includes('compiler-phase-benchmark')
) {
  const verbose = process.argv.includes('--verbose');
  const reduced = process.argv.includes('--reduced');
  const config = reduced ? REDUCED_CONFIG : DEFAULT_CONFIG;

  runCompilerPhaseBenchmark(config).then((report) => {
    if (verbose) {
      console.log(JSON.stringify(report, null, 2));
    } else {
      console.log(`\n${'═'.repeat(72)}`);
      console.log('  Gnosis Compiler Family Benchmark');
      console.log(`${'═'.repeat(72)}\n`);

      // Compiler shootout
      console.log('── Compiler Shootout: Betty vs Betti vs aeon-logic ──\n');
      for (const topo of report.compilerShootout) {
        console.log(`  ${topo.topologyName} (${topo.lineCount} lines)`);
        for (const c of [...topo.compilers].sort(
          (a, b) => a.meanMs - b.meanMs
        )) {
          const tag = c.compiler === topo.winner ? ' *' : '';
          const bar = '█'.repeat(Math.max(1, Math.round(c.meanMs * 10)));
          console.log(
            `    ${c.compiler.padEnd(14)} ${c.meanMs
              .toFixed(3)
              .padStart(8)}ms  (${c.nodeCount}N ${c.edgeCount}E b1=${
              c.beta1
            })  ${bar}${tag}`
          );
        }
        console.log(
          `    winner: ${topo.winner} (${topo.winnerSpeedup.toFixed(
            1
          )}x faster than slowest)\n`
        );
      }

      // Compiler ranking
      console.log('── Compiler Ranking (averaged across all topologies) ──');
      for (const c of report.headToHead.compilerRanking) {
        const pct = (c.fractionOfSlowest * 100).toFixed(1);
        console.log(
          `  ${c.compiler.padEnd(14)} ${c.meanMs
            .toFixed(3)
            .padStart(8)}ms  (${pct}% of slowest)`
        );
      }

      // Phase breakdown
      console.log('\n── Betty Phase Breakdown (by size class) ──\n');
      for (const sc of report.sizeClasses) {
        console.log(
          `  ${sc.sizeClass.label} (${sc.sizeClass.meanLineCount.toFixed(
            0
          )} lines, ${sc.sizeClass.meanNodeCount.toFixed(
            0
          )} nodes, ${sc.sizeClass.meanEdgeCount.toFixed(0)} edges)`
        );
        for (const phase of sc.aggregatePhases) {
          if (phase.timingsMs.length === 0) continue;
          const bar = '█'.repeat(Math.max(1, Math.round(phase.meanMs * 10)));
          console.log(
            `    ${phase.phase.padEnd(24)} ${phase.meanMs
              .toFixed(3)
              .padStart(8)}ms  ${bar}`
          );
        }
        console.log();
      }

      console.log('── Head-to-Head: aeon-logic vs Betty ──');
      for (const h of report.headToHead.aeonLogicVsBetty) {
        console.log(
          `  ${h.sizeClass.padEnd(10)} aeon-logic: ${h.aeonLogicMeanMs.toFixed(
            3
          )}ms  betty: ${h.bettyFullMeanMs.toFixed(
            3
          )}ms  overhead: ${h.overhead.toFixed(1)}x`
        );
      }

      console.log('\n── Phase Cost Ranking (all sizes) ──');
      for (const p of report.headToHead.phaseCostRanking) {
        if (p.meanMs === 0) continue;
        console.log(
          `  ${p.phase.padEnd(24)} ${p.meanMs.toFixed(3).padStart(8)}ms  (${(
            p.fractionOfTotal * 100
          ).toFixed(1)}%)`
        );
      }
    }
  });
}
