// @a0n/gnosis/src/benchmarks/foil-admission-shootout.ts
import {
  createStaticFoilAdmissionProof,
  FoilAdmissionRejectedError,
  FoilEdgeRuntime,
  type FoilEdgeAdmissionMetrics,
  type ForkRaceFoldCertificateSummary,
  type TopologyPayload,
  ZERO_DEFICIT_FOIL_CERTIFICATE,
} from '../foil-edge-gate';
import { FoilEHLARuntime } from '../foil-ehla-runtime';

export interface FoilAdmissionShootoutScenario {
  readonly id: string;
  readonly payload: TopologyPayload;
}

export interface FoilAdmissionShootoutReport {
  readonly label: 'gnosis-foil-admission-shootout-v1';
  readonly leanTheorems: readonly string[];
  readonly scenarioCount: number;
  readonly acceptedByBaseline: number;
  readonly acceptedByTritonGate: number;
  readonly acceptedByEHLA: number;
  readonly baselineMetrics: FoilEdgeAdmissionMetrics;
  readonly tritonGateMetrics: FoilEdgeAdmissionMetrics;
  readonly ehlaMetrics: FoilEdgeAdmissionMetrics & { 
    readonly interpretationDebt: number;
    readonly harvestedTopologies: number;
  };
  readonly baselineVerifierCalls: number;
  readonly tritonGateVerifierCalls: number;
  readonly ehlaVerifierCalls: number;
  readonly avoidedVerifierCalls: number;
  readonly ehlaAvoidedVerifierCalls: number;
  readonly avoidedVerificationCost: number;
  readonly escapeHatchVerifierCalls: number;
}

/**
 * Handles the gnosis run Foil Admission Shootout workflow.
 */
export function runFoilAdmissionShootout(
  scenarios: readonly FoilAdmissionShootoutScenario[] = defaultFoilAdmissionShootoutScenarios()
): FoilAdmissionShootoutReport {
  const baselineRuntime = new FoilEdgeRuntime();
  const tritonGateRuntime = new FoilEdgeRuntime();
  const ehlaRuntime = new FoilEHLARuntime({ maxDebt: 10 });

  let acceptedByBaseline = 0;
  let acceptedByTritonGate = 0;
  let acceptedByEHLA = 0;
  let baselineVerifierCost = 0;
  let tritonGateVerifierCost = 0;

  for (const scenario of scenarios) {
    const baselinePayload = stripTritonAdmission(scenario.payload);
    try {
      baselineRuntime.executeTopology(baselinePayload, () => undefined);
      acceptedByBaseline += 1;
      baselineVerifierCost += baselinePayload.zkProof.verificationCost;
    } catch (error) {
      if (!(error instanceof FoilAdmissionRejectedError)) {
        throw error;
      }
    }

    try {
      tritonGateRuntime.executeTopology(scenario.payload, () => undefined);
      acceptedByTritonGate += 1;
      tritonGateVerifierCost += scenario.payload.zkProof.verificationCost;
    } catch (error) {
      if (!(error instanceof FoilAdmissionRejectedError)) {
        throw error;
      }
    }

    try {
      ehlaRuntime.executeTopologyEHLA(scenario.payload, () => undefined);
      acceptedByEHLA += 1;
    } catch (error) {
      if (!(error instanceof FoilAdmissionRejectedError)) {
        throw error;
      }
    }
  }

  const baselineMetrics = baselineRuntime.metrics();
  const tritonGateMetrics = tritonGateRuntime.metrics();
  const ehlaStats = ehlaRuntime.getEHLAStats();
  const avoidedVerifierCalls =
    baselineMetrics.verifierCalls - tritonGateMetrics.verifierCalls;

  return {
    label: 'gnosis-foil-admission-shootout-v1',
    leanTheorems: [
      'Gnosis.PleromaAeonMonsterBridge.weighted_triton_verifier_cost_le_baseline',
      'Gnosis.PleromaAeonMonsterBridge.weighted_triton_avoided_cost_eq_dropped_cost',
      'Gnosis.PleromaAeonMonsterBridge.weighted_foil_admission_shootout_avoided_cost',
      'Gnosis.EntropyHarvestingArchitecture.ehla_harvests_debt_as_propellant',
    ],
    scenarioCount: scenarios.length,
    acceptedByBaseline,
    acceptedByTritonGate,
    acceptedByEHLA,
    baselineMetrics,
    tritonGateMetrics,
    ehlaMetrics: ehlaStats,
    baselineVerifierCalls: baselineMetrics.verifierCalls,
    tritonGateVerifierCalls: tritonGateMetrics.verifierCalls,
    ehlaVerifierCalls: ehlaStats.verifierCalls,
    avoidedVerifierCalls,
    ehlaAvoidedVerifierCalls: baselineMetrics.verifierCalls - ehlaStats.verifierCalls,
    avoidedVerificationCost: baselineVerifierCost - tritonGateVerifierCost,
    escapeHatchVerifierCalls:
      tritonGateMetrics.verifierCalls - tritonGateMetrics.tritonPreflightAdmitted,
  };
}

/**
 * Handles the gnosis default Foil Admission Shootout Scenarios workflow.
 */
export function defaultFoilAdmissionShootoutScenarios(): FoilAdmissionShootoutScenario[] {
  return [
    {
      id: 'valid-race-fold',
      payload: {
        ggGraph: new Uint8Array([1]),
        zkProof: createProof(ZERO_DEFICIT_FOIL_CERTIFICATE, 5),
        tritonAdmission: {
          phase: 'race-fold-present',
          clockPhase: 0,
          carrierKey: 11,
        },
      },
    },
    {
      id: 'wrong-clock-slot',
      payload: {
        ggGraph: new Uint8Array([2]),
        zkProof: createProof(ZERO_DEFICIT_FOIL_CERTIFICATE, 7),
        tritonAdmission: {
          phase: 'fork-future',
          clockPhase: 0,
        },
      },
    },
    {
      id: 'fork-budget-exhausted',
      payload: {
        ggGraph: new Uint8Array([3]),
        zkProof: createProof(ZERO_DEFICIT_FOIL_CERTIFICATE, 11),
        tritonAdmission: {
          phase: 'fork-future',
          clockPhase: 1,
          forkBudget: 0,
        },
      },
    },
    {
      id: 'c2-duplicate',
      payload: {
        ggGraph: new Uint8Array([4]),
        zkProof: createProof(ZERO_DEFICIT_FOIL_CERTIFICATE, 13),
        tritonAdmission: {
          phase: 'race-fold-present',
          clockPhase: 0,
          carrierKey: 1,
          seenCanonicalC2Keys: new Set([0]),
        },
      },
    },
    {
      id: 'escape-hatch-no-triton',
      payload: {
        ggGraph: new Uint8Array([5]),
        zkProof: createProof(ZERO_DEFICIT_FOIL_CERTIFICATE, 17),
      },
    },
  ];
}

function createProof(
  certificate: ForkRaceFoldCertificateSummary,
  verificationCost: number
) {
  return createStaticFoilAdmissionProof(certificate, {
    proofBytes: 1,
    verificationCost,
    constantBound: verificationCost,
  });
}

function stripTritonAdmission(payload: TopologyPayload): TopologyPayload {
  return {
    ggGraph: payload.ggGraph,
    zkProof: payload.zkProof,
  };
}
