// @a0n/gnosis/src/benchmarks/moa-transformer-evidence-benchmark.ts
import {
  DEFAULT_MOA_TRANSFORMER_TOPOLOGY_FILES,
  makeDefaultMoaTransformerShootoutConfig,
  runGnosisMoaTransformerFamilyBenchmark,
  runGnosisMoaTransformerShootoutBenchmark,
  type MoaTransformerFamilyReport,
  type GnosisMoaTransformerShootoutReport,
  type MoaTransformerFamily,
  type MoaTransformerShootoutConfig,
} from './moa-transformer-shootout-benchmark';

export interface MoaTransformerEvidenceScaleConfig {
  readonly id: string;
  readonly title: string;
  readonly trainAxisPointCount: number;
  readonly evalAxisPointCount: number;
  readonly epochsMultiplier: number;
}

export interface MoaTransformerEvidenceAblationConfig {
  readonly id: string;
  readonly title: string;
  readonly description: string;
  readonly activeBlockCount: number;
  readonly activeHeadCount: number;
}

export interface MoaTransformerEvidenceConfig {
  readonly label: 'gnosis-moa-transformer-evidence-v1';
  readonly base: MoaTransformerShootoutConfig;
  readonly scales: readonly MoaTransformerEvidenceScaleConfig[];
  readonly ablations: readonly MoaTransformerEvidenceAblationConfig[];
}

export interface MoaTransformerEvidenceFamilySummary {
  readonly topologyPath: string;
  readonly meanEvalMeanSquaredError: number;
  readonly meanExactWithinToleranceFraction: number;
  readonly meanEvalWallTimeMs: number;
  readonly meanEvalSamplesPerSecond: number;
  readonly meanActiveBlockCount: number;
  readonly meanActiveHeadCount: number;
  readonly meanFrameCount: number;
  readonly computeAdjustedExactFraction: number;
}

export interface MoaTransformerEvidenceScaleReport {
  readonly id: string;
  readonly title: string;
  readonly epochs: number;
  readonly trainSampleCount: number;
  readonly evalSampleCount: number;
  readonly families: Record<
    MoaTransformerFamily,
    MoaTransformerEvidenceFamilySummary
  >;
  readonly moaEvalWallTimeSpeedupVsRegular: number;
  readonly moaHeadReductionFactor: number;
  readonly moaFrameReductionFactor: number;
  readonly rawAccuracyCompetitive: boolean;
  readonly moaUsesLessCompute: boolean;
}

export interface MoaTransformerEvidenceAblationReport {
  readonly id: string;
  readonly title: string;
  readonly description: string;
  readonly activeBlockCount: number;
  readonly activeHeadCount: number;
  readonly families: Record<
    MoaTransformerFamily,
    MoaTransformerEvidenceFamilySummary
  >;
  readonly moaEvalWallTimeSpeedupVsRegular: number;
  readonly rawAccuracyCompetitive: boolean;
  readonly moaUsesLessCompute: boolean;
}

export interface GnosisMoaTransformerEvidenceReport {
  readonly label: MoaTransformerEvidenceConfig['label'];
  readonly topologySurface: {
    readonly regularTopologyPath: string;
    readonly moaTopologyPath: string;
    readonly moaStructuredPrimitive: 'StructuredMoA';
  };
  readonly baseConfig: {
    readonly seedCount: number;
    readonly learningRate: number;
    readonly exactPredictionTolerance: number;
    readonly dualActiveThreshold: number;
    readonly frameStageCount: number;
    readonly samplePoints: readonly (readonly [number, number])[];
  };
  readonly scales: readonly MoaTransformerEvidenceScaleReport[];
  readonly ablations: readonly MoaTransformerEvidenceAblationReport[];
  readonly timingAdvantageRecovered: boolean;
  readonly accuracyGapClosesWithScale: boolean;
  readonly outerSparsityImprovesEfficiency: boolean;
  readonly innerSparsityImprovesEfficiency: boolean;
  readonly underRoutingHurtsAccuracy: boolean;
}

function makeAxis(pointCount: number, maxAbs: number): readonly number[] {
  if (pointCount <= 1) {
    return [0];
  }

  return Array.from({ length: pointCount }, (_, index) => {
    const fraction = index / (pointCount - 1);
    const value = -maxAbs + 2 * maxAbs * fraction;
    return Number(value.toFixed(3));
  });
}

function summarizeFamily(
  report: GnosisMoaTransformerShootoutReport,
  family: MoaTransformerFamily
): MoaTransformerEvidenceFamilySummary {
  const metrics = report.families[family];
  return {
    topologyPath: metrics.topologyPath,
    meanEvalMeanSquaredError: metrics.meanEvalMeanSquaredError,
    meanExactWithinToleranceFraction: metrics.meanExactWithinToleranceFraction,
    meanEvalWallTimeMs: metrics.meanEvalWallTimeMs,
    meanEvalSamplesPerSecond: metrics.meanEvalSamplesPerSecond,
    meanActiveBlockCount: metrics.meanActiveBlockCount,
    meanActiveHeadCount: metrics.meanActiveHeadCount,
    meanFrameCount: metrics.meanFrameCount,
    computeAdjustedExactFraction: metrics.computeAdjustedExactFraction,
  };
}

function summarizeScale(
  scale: MoaTransformerEvidenceScaleConfig,
  report: GnosisMoaTransformerShootoutReport
): MoaTransformerEvidenceScaleReport {
  const regular = summarizeFamily(report, 'regular');
  const moa = summarizeFamily(report, 'moa');
  return {
    id: scale.id,
    title: scale.title,
    epochs: report.config.epochs,
    trainSampleCount: report.config.trainSampleCount,
    evalSampleCount: report.config.evalSampleCount,
    families: { regular, moa },
    moaEvalWallTimeSpeedupVsRegular:
      moa.meanEvalWallTimeMs <= 0
        ? 0
        : regular.meanEvalWallTimeMs / moa.meanEvalWallTimeMs,
    moaHeadReductionFactor:
      moa.meanActiveHeadCount <= 0
        ? 0
        : regular.meanActiveHeadCount / moa.meanActiveHeadCount,
    moaFrameReductionFactor:
      moa.meanFrameCount <= 0 ? 0 : regular.meanFrameCount / moa.meanFrameCount,
    rawAccuracyCompetitive: report.rawAccuracyCompetitive,
    moaUsesLessCompute: report.moaUsesLessCompute,
  };
}

function summarizeAblation(
  ablation: MoaTransformerEvidenceAblationConfig,
  regular: MoaTransformerFamilyReport,
  moa: MoaTransformerFamilyReport
): MoaTransformerEvidenceAblationReport {
  return {
    id: ablation.id,
    title: ablation.title,
    description: ablation.description,
    activeBlockCount: ablation.activeBlockCount,
    activeHeadCount: ablation.activeHeadCount,
    families: { regular, moa },
    moaEvalWallTimeSpeedupVsRegular:
      moa.meanEvalWallTimeMs <= 0
        ? 0
        : regular.meanEvalWallTimeMs / moa.meanEvalWallTimeMs,
    rawAccuracyCompetitive:
      moa.meanEvalMeanSquaredError <= regular.meanEvalMeanSquaredError + 0.03 &&
      moa.meanExactWithinToleranceFraction >=
        regular.meanExactWithinToleranceFraction - 0.05,
    moaUsesLessCompute:
      moa.meanActiveHeadCount < regular.meanActiveHeadCount &&
      moa.meanFrameCount < regular.meanFrameCount,
  };
}

function usesBaseMoaRouting(
  ablation: MoaTransformerEvidenceAblationConfig,
  base: MoaTransformerShootoutConfig
): boolean {
  return (
    ablation.activeBlockCount === base.activeBlockCount &&
    ablation.activeHeadCount === base.activeHeadCount
  );
}

function hasPrimarySweepTimingAdvantage(
  scales: readonly MoaTransformerEvidenceScaleReport[]
): boolean {
  if (scales.length === 0) {
    return false;
  }

  const timingWins = scales.filter(
    (scale) => scale.moaEvalWallTimeSpeedupVsRegular > 1
  ).length;
  const requiredWins = Math.ceil((scales.length * 2) / 3);

  return (
    timingWins >= requiredWins &&
    scales.every((scale) => scale.moaUsesLessCompute)
  );
}

/**
 * Handles the gnosis make Default Moa Transformer Evidence Config workflow.
 */
export function makeDefaultMoaTransformerEvidenceConfig(): MoaTransformerEvidenceConfig {
  return {
    label: 'gnosis-moa-transformer-evidence-v1',
    base: {
      ...makeDefaultMoaTransformerShootoutConfig(),
      seedCount: 4,
      epochs: 500,
    },
    scales: [
      {
        id: 'compact',
        title: 'Compact workload',
        trainAxisPointCount: 5,
        evalAxisPointCount: 4,
        epochsMultiplier: 0.8,
      },
      {
        id: 'baseline',
        title: 'Baseline workload',
        trainAxisPointCount: 7,
        evalAxisPointCount: 5,
        epochsMultiplier: 1,
      },
      {
        id: 'wide',
        title: 'Wide workload',
        trainAxisPointCount: 9,
        evalAxisPointCount: 7,
        epochsMultiplier: 1.2,
      },
    ],
    ablations: [
      {
        id: 'full-moa',
        title: 'Full MoA',
        description:
          'Outer sparsity and inner head sparsity both active at the paper baseline.',
        activeBlockCount: 2,
        activeHeadCount: 2,
      },
      {
        id: 'no-outer-sparsity',
        title: 'No outer sparsity',
        description:
          'Keep all four blocks live while preserving the sparse inner head whip.',
        activeBlockCount: 4,
        activeHeadCount: 2,
      },
      {
        id: 'no-inner-sparsity',
        title: 'No inner sparsity',
        description:
          'Keep the outer MoA route but expand each live block to all four heads.',
        activeBlockCount: 2,
        activeHeadCount: 4,
      },
      {
        id: 'under-routed',
        title: 'Under-routed',
        description:
          'Over-sparsify the outer route to a single block and observe the accuracy penalty.',
        activeBlockCount: 1,
        activeHeadCount: 2,
      },
    ],
  };
}

/**
 * Handles the gnosis run Gnosis Moa Transformer Evidence Benchmark workflow.
 */
export async function runGnosisMoaTransformerEvidenceBenchmark(
  config: MoaTransformerEvidenceConfig = makeDefaultMoaTransformerEvidenceConfig()
): Promise<GnosisMoaTransformerEvidenceReport> {
  const scales: MoaTransformerEvidenceScaleReport[] = [];
  for (const scale of config.scales) {
    const epochs = Math.max(
      1,
      Math.round(config.base.epochs * scale.epochsMultiplier)
    );
    const report = await runGnosisMoaTransformerShootoutBenchmark({
      ...config.base,
      epochs,
      trainAxis: makeAxis(scale.trainAxisPointCount, 1.5),
      evalAxis: makeAxis(scale.evalAxisPointCount, 1.4),
      samplePoints: [],
    });
    scales.push(summarizeScale(scale, report));
  }

  const baselineAblationReport = await runGnosisMoaTransformerShootoutBenchmark(
    {
      ...config.base,
      samplePoints: [],
    }
  );
  const baselineRegular = baselineAblationReport.families.regular;

  const ablations: MoaTransformerEvidenceAblationReport[] = [];
  for (const ablation of config.ablations) {
    const moaFamily = usesBaseMoaRouting(ablation, config.base)
      ? baselineAblationReport.families.moa
      : await runGnosisMoaTransformerFamilyBenchmark('moa', {
          ...config.base,
          activeBlockCount: ablation.activeBlockCount,
          activeHeadCount: ablation.activeHeadCount,
          samplePoints: [],
        });
    ablations.push(summarizeAblation(ablation, baselineRegular, moaFamily));
  }

  const fullMoa = ablations.find((entry) => entry.id === 'full-moa');
  const noOuter = ablations.find((entry) => entry.id === 'no-outer-sparsity');
  const noInner = ablations.find((entry) => entry.id === 'no-inner-sparsity');
  const underRouted = ablations.find((entry) => entry.id === 'under-routed');
  const scaleGaps = scales.map(
    (scale) =>
      scale.families.moa.meanEvalMeanSquaredError -
      scale.families.regular.meanEvalMeanSquaredError
  );

  return {
    label: config.label,
    topologySurface: {
      regularTopologyPath: DEFAULT_MOA_TRANSFORMER_TOPOLOGY_FILES.regular,
      moaTopologyPath: DEFAULT_MOA_TRANSFORMER_TOPOLOGY_FILES.moa,
      moaStructuredPrimitive: 'StructuredMoA',
    },
    baseConfig: {
      seedCount: config.base.seedCount,
      learningRate: config.base.learningRate,
      exactPredictionTolerance: config.base.exactPredictionTolerance,
      dualActiveThreshold: config.base.dualActiveThreshold,
      frameStageCount: config.base.frameStageCount,
      samplePoints: config.base.samplePoints,
    },
    scales,
    ablations,
    timingAdvantageRecovered: hasPrimarySweepTimingAdvantage(scales),
    accuracyGapClosesWithScale:
      scaleGaps.length <= 1 ||
      ((scaleGaps[scaleGaps.length - 1] ?? 0) < (scaleGaps[0] ?? 0) &&
        scaleGaps.every(
          (gap, index) => index === 0 || gap <= (scaleGaps[index - 1] ?? gap)
        )),
    outerSparsityImprovesEfficiency:
      fullMoa !== undefined &&
      noOuter !== undefined &&
      fullMoa.families.moa.computeAdjustedExactFraction >
        noOuter.families.moa.computeAdjustedExactFraction &&
      fullMoa.families.moa.meanActiveBlockCount <
        noOuter.families.moa.meanActiveBlockCount,
    innerSparsityImprovesEfficiency:
      fullMoa !== undefined &&
      noInner !== undefined &&
      fullMoa.families.moa.computeAdjustedExactFraction >
        noInner.families.moa.computeAdjustedExactFraction &&
      fullMoa.families.moa.meanActiveHeadCount <
        noInner.families.moa.meanActiveHeadCount,
    underRoutingHurtsAccuracy:
      fullMoa !== undefined &&
      underRouted !== undefined &&
      underRouted.families.moa.meanEvalMeanSquaredError >
        fullMoa.families.moa.meanEvalMeanSquaredError &&
      underRouted.families.moa.meanExactWithinToleranceFraction <
        fullMoa.families.moa.meanExactWithinToleranceFraction,
  };
}

/**
 * Handles the gnosis render Gnosis Moa Transformer Evidence Markdown workflow.
 */
export function renderGnosisMoaTransformerEvidenceMarkdown(
  report: GnosisMoaTransformerEvidenceReport
): string {
  const lines: string[] = [];
  lines.push('# Gnosis MoA Transformer Evidence');
  lines.push('');
  lines.push(`- Label: \`${report.label}\``);
  lines.push(
    `- Regular GG topology: \`${report.topologySurface.regularTopologyPath}\``
  );
  lines.push(
    `- Sparse GG topology: \`${report.topologySurface.moaTopologyPath}\``
  );
  lines.push(
    `- Sparse GG primitive: \`${report.topologySurface.moaStructuredPrimitive}\``
  );
  lines.push(`- Base seed count: \`${report.baseConfig.seedCount}\``);
  lines.push(`- Learning rate: \`${report.baseConfig.learningRate}\``);
  lines.push(
    `- Timing advantage recovered across sweep: \`${
      report.timingAdvantageRecovered ? 'yes' : 'no'
    }\``
  );
  lines.push(
    `- Accuracy gap closes with scale: \`${
      report.accuracyGapClosesWithScale ? 'yes' : 'no'
    }\``
  );
  lines.push(
    `- Outer sparsity improves efficiency: \`${
      report.outerSparsityImprovesEfficiency ? 'yes' : 'no'
    }\``
  );
  lines.push(
    `- Inner sparsity improves efficiency: \`${
      report.innerSparsityImprovesEfficiency ? 'yes' : 'no'
    }\``
  );
  lines.push(
    `- Under-routing hurts accuracy: \`${
      report.underRoutingHurtsAccuracy ? 'yes' : 'no'
    }\``
  );
  lines.push('');
  lines.push('## Scale Sweep');
  lines.push('');
  lines.push(
    '| Scale | Train samples | Eval samples | MoA eval MSE | Regular eval MSE | MoA eval speedup | MoA active heads | Regular active heads | MoA frames | Regular frames |'
  );
  lines.push('|---|---:|---:|---:|---:|---:|---:|---:|---:|---:|');
  for (const scale of report.scales) {
    lines.push(
      `| ${scale.title} | ${scale.trainSampleCount} | ${
        scale.evalSampleCount
      } | ${scale.families.moa.meanEvalMeanSquaredError.toFixed(
        4
      )} | ${scale.families.regular.meanEvalMeanSquaredError.toFixed(
        4
      )} | ${scale.moaEvalWallTimeSpeedupVsRegular.toFixed(
        2
      )}x | ${scale.families.moa.meanActiveHeadCount.toFixed(
        1
      )} | ${scale.families.regular.meanActiveHeadCount.toFixed(
        1
      )} | ${scale.families.moa.meanFrameCount.toFixed(
        1
      )} | ${scale.families.regular.meanFrameCount.toFixed(1)} |`
    );
  }

  lines.push('');
  lines.push('## Ablations');
  lines.push('');
  lines.push(
    '| Ablation | Active blocks | Active heads | MoA eval MSE | MoA exact fraction | MoA eval wall ms | Compute-adjusted exact |'
  );
  lines.push('|---|---:|---:|---:|---:|---:|---:|');
  for (const ablation of report.ablations) {
    lines.push(
      `| ${ablation.title} | ${ablation.activeBlockCount} | ${
        ablation.activeHeadCount
      } | ${ablation.families.moa.meanEvalMeanSquaredError.toFixed(
        4
      )} | ${ablation.families.moa.meanExactWithinToleranceFraction.toFixed(
        4
      )} | ${ablation.families.moa.meanEvalWallTimeMs.toFixed(
        2
      )} | ${ablation.families.moa.computeAdjustedExactFraction.toFixed(4)} |`
    );
  }

  lines.push('');
  lines.push(
    'Interpretation: the same lower-triangular fork/race/fold geometry now reappears across head chains, transformer blocks, and runtime structured concurrency. The sparse family is declared in GG through `StructuredMoA`, then lowered into the same nested Rotation-plus-Whip graph that the benchmark executes. The evidence pass checks whether that recursive Wallington/MoA design keeps its accuracy while genuinely reducing live compute, and whether the outer route and inner head sparsity are both doing real work rather than decorative pruning.'
  );
  lines.push('');
  return `${lines.join('\n')}\n`;
}
