// @a0n/gnosis/src/benchmarks/moa-transformer-shootout-benchmark.ts
import { readFile } from 'node:fs/promises';
import { fileURLToPath } from 'node:url';

import {
  FIN,
  FOLD,
  FORK,
  FlowCodec,
  FrameReassembler,
  RACE,
  type FlowFrame,
} from '@a0n/aeon/flow';
import { parseGgProgram } from '@a0n/aeon-logic';
import { lowerUfcsSource } from '../ufcs';
import {
  bootstrapMeanConfidenceInterval,
  createDeterministicRandom,
  mean,
  stdev,
  type ConfidenceInterval,
} from './statistics';

export type MoaTransformerFamily = 'regular' | 'moa';

export interface MoaTransformerShootoutConfig {
  readonly label: 'gnosis-moa-transformer-shootout-v1';
  readonly seedCount: number;
  readonly epochs: number;
  readonly learningRate: number;
  readonly exactPredictionTolerance: number;
  readonly dualActiveThreshold: number;
  readonly frameStageCount: number;
  readonly activeBlockCount: number;
  readonly activeHeadCount: number;
  readonly trainAxis: readonly number[];
  readonly evalAxis: readonly number[];
  readonly samplePoints: readonly (readonly [number, number])[];
}

export interface MoaTransformerSeedMetrics {
  readonly seed: number;
  readonly trainMeanSquaredError: number;
  readonly evalMeanSquaredError: number;
  readonly exactWithinToleranceFraction: number;
  readonly dualActiveRegionMeanAbsoluteError: number;
  readonly trainWallTimeMs: number;
  readonly evalWallTimeMs: number;
  readonly evalSamplesPerSecond: number;
  readonly codecRoundTripExactFraction: number;
  readonly reassemblyExactFraction: number;
  readonly foldInvarianceExactFraction: number;
  readonly activeBlockCount: number;
  readonly activeHeadCount: number;
  readonly frameCount: number;
  readonly payloadBytes: number;
}

export interface MoaTransformerSamplePrediction {
  readonly leftInput: number;
  readonly rightInput: number;
  readonly target: number;
  readonly meanPrediction: number;
  readonly predictionStddev: number;
  readonly meanActiveBlockCount: number;
  readonly meanActiveHeadCount: number;
  readonly meanFrameCount: number;
}

export interface MoaTransformerTopologyMetrics {
  readonly nodeCount: number;
  readonly edgeCount: number;
  readonly structuralBeta1: number;
  readonly transformerletCount: number;
  readonly headChainCount: number;
  readonly rotationStageCount: number;
  readonly parameterCount: number;
  readonly usesOuterRouter: boolean;
}

export interface MoaTransformerFamilyReport {
  readonly topologyPath: string;
  readonly topologyNodeCount: number;
  readonly topologyEdgeCount: number;
  readonly structuralBeta1: number;
  readonly transformerletCount: number;
  readonly headChainCount: number;
  readonly rotationStageCount: number;
  readonly parameterCount: number;
  readonly usesOuterRouter: boolean;
  readonly seedMetrics: readonly MoaTransformerSeedMetrics[];
  readonly meanTrainMeanSquaredError: number;
  readonly stdevTrainMeanSquaredError: number;
  readonly meanEvalMeanSquaredError: number;
  readonly evalMeanSquaredErrorCi95: ConfidenceInterval;
  readonly stdevEvalMeanSquaredError: number;
  readonly meanExactWithinToleranceFraction: number;
  readonly exactWithinToleranceFractionCi95: ConfidenceInterval;
  readonly stdevExactWithinToleranceFraction: number;
  readonly meanDualActiveRegionMeanAbsoluteError: number;
  readonly dualActiveRegionMeanAbsoluteErrorCi95: ConfidenceInterval;
  readonly stdevDualActiveRegionMeanAbsoluteError: number;
  readonly meanTrainWallTimeMs: number;
  readonly meanEvalWallTimeMs: number;
  readonly meanEvalSamplesPerSecond: number;
  readonly meanCodecRoundTripExactFraction: number;
  readonly meanReassemblyExactFraction: number;
  readonly meanFoldInvarianceExactFraction: number;
  readonly meanActiveBlockCount: number;
  readonly meanActiveHeadCount: number;
  readonly meanHeadUtilizationFraction: number;
  readonly meanFrameCount: number;
  readonly meanPayloadBytes: number;
  readonly computeAdjustedExactFraction: number;
  readonly computeAdjustedEvalLoss: number;
  readonly samplePredictions: readonly MoaTransformerSamplePrediction[];
}

export interface GnosisMoaTransformerShootoutReport {
  readonly label: MoaTransformerShootoutConfig['label'];
  readonly config: {
    readonly seedCount: number;
    readonly epochs: number;
    readonly learningRate: number;
    readonly exactPredictionTolerance: number;
    readonly dualActiveThreshold: number;
    readonly frameStageCount: number;
    readonly activeBlockCount: number;
    readonly activeHeadCount: number;
    readonly trainSampleCount: number;
    readonly evalSampleCount: number;
    readonly targetFamily: 'quadrant-l1-rotated-transformer';
  };
  readonly families: Record<MoaTransformerFamily, MoaTransformerFamilyReport>;
  readonly sharedCapacity: {
    readonly transformerletCount: number;
    readonly headChainCount: number;
    readonly rotationStageCount: number;
    readonly parameterCount: number;
  };
  readonly winnerByEvalMeanSquaredError: MoaTransformerFamily;
  readonly winnerByComputeAdjustedExactFraction: MoaTransformerFamily;
  readonly moaUsesLessCompute: boolean;
  readonly rawAccuracyCompetitive: boolean;
}

interface Sample {
  readonly leftInput: number;
  readonly rightInput: number;
  readonly target: number;
}

interface HeadParameters {
  gateWeight: number;
  gateBias: number;
  valueWeight: number;
  valueBias: number;
}

interface BlockParameters {
  outerGateWeight: number;
  outerGateBias: number;
  heads: HeadParameters[];
}

interface ParsedTopology {
  readonly family: MoaTransformerFamily;
  readonly path: string;
  readonly metrics: MoaTransformerTopologyMetrics;
}

let defaultTopologySuitePromise:
  | Promise<Record<MoaTransformerFamily, ParsedTopology>>
  | undefined;

interface BlockSpec {
  readonly id: string;
  readonly streamBaseId: number;
  readonly signedProjection: (sample: Sample) => number;
  readonly axisProjection: (sample: Sample) => number;
}

interface HeadSpec {
  readonly id: string;
  readonly magnitudeCenter: number;
  readonly inputScale: number;
  readonly valueScale: number;
}

interface HeadActivation {
  readonly gate: number;
  readonly value: number;
  readonly output: number;
  readonly headInput: number;
  readonly valueInput: number;
  readonly selected: boolean;
}

interface BlockActivation {
  readonly outerGate: number;
  readonly blockOutput: number;
  readonly signedInput: number;
  readonly axisInput: number;
  readonly active: boolean;
  readonly selectedHeadCount: number;
  readonly headActivations: readonly HeadActivation[];
}

interface PredictionResult {
  readonly prediction: number;
  readonly activeBlockCount: number;
  readonly activeHeadCount: number;
  readonly blockActivations: readonly BlockActivation[];
}

interface EvaluationMetrics {
  readonly meanSquaredError: number;
  readonly exactWithinToleranceFraction: number;
  readonly dualActiveRegionMeanAbsoluteError: number;
  readonly activeBlockCount: number;
  readonly activeHeadCount: number;
}

interface TransportMetrics {
  readonly codecRoundTripExactFraction: number;
  readonly reassemblyExactFraction: number;
  readonly foldInvarianceExactFraction: number;
  readonly frameCount: number;
  readonly payloadBytes: number;
}

interface FramedPredictionResult extends TransportMetrics {
  readonly prediction: number;
  readonly activeBlockCount: number;
  readonly activeHeadCount: number;
}

function createDormantHeadActivation(): HeadActivation {
  return {
    gate: 0,
    value: 0,
    output: 0,
    headInput: 0,
    valueInput: 0,
    selected: false,
  };
}

const DEFAULT_MOA_TRANSFORMER_TOPOLOGY_URLS: Record<MoaTransformerFamily, URL> =
  {
    regular: new URL(
      '../../examples/benchmarks/moa-transformer-regular.gg',
      import.meta.url
    ),
    moa: new URL(
      '../../examples/benchmarks/moa-transformer-moa.gg',
      import.meta.url
    ),
  };

export const DEFAULT_MOA_TRANSFORMER_TOPOLOGY_FILES = Object.fromEntries(
  (
    Object.entries(DEFAULT_MOA_TRANSFORMER_TOPOLOGY_URLS) as Array<
      readonly [MoaTransformerFamily, URL]
    >
  ).map(([family, url]) => [family, fileURLToPath(url)])
) as Record<MoaTransformerFamily, string>;

const blockSpecs: readonly BlockSpec[] = [
  {
    id: 'x_pos',
    streamBaseId: 1,
    signedProjection: (sample) => sample.leftInput,
    axisProjection: (sample) => sample.leftInput,
  },
  {
    id: 'x_neg',
    streamBaseId: 2,
    signedProjection: (sample) => -sample.leftInput,
    axisProjection: (sample) => sample.leftInput,
  },
  {
    id: 'y_pos',
    streamBaseId: 3,
    signedProjection: (sample) => sample.rightInput,
    axisProjection: (sample) => sample.rightInput,
  },
  {
    id: 'y_neg',
    streamBaseId: 4,
    signedProjection: (sample) => -sample.rightInput,
    axisProjection: (sample) => sample.rightInput,
  },
] as const;

const headSpecs: readonly HeadSpec[] = [
  { id: 'h0', magnitudeCenter: 0.15, inputScale: 0.6, valueScale: 0.5 },
  { id: 'h1', magnitudeCenter: 0.45, inputScale: 0.9, valueScale: 0.8 },
  { id: 'h2', magnitudeCenter: 0.9, inputScale: 1.2, valueScale: 1.1 },
  { id: 'h3', magnitudeCenter: 1.35, inputScale: 1.5, valueScale: 1.4 },
] as const;

function sigmoid(value: number): number {
  return 1 / (1 + Math.exp(-value));
}

function countConnectedComponents(
  nodeIds: readonly string[],
  edges: readonly {
    sourceIds: readonly string[];
    targetIds: readonly string[];
  }[]
): number {
  const neighbors = new Map<string, Set<string>>();
  for (const nodeId of nodeIds) {
    neighbors.set(nodeId, new Set<string>());
  }

  for (const edge of edges) {
    for (const sourceId of edge.sourceIds) {
      const sourceNeighbors = neighbors.get(sourceId);
      if (!sourceNeighbors) {
        continue;
      }
      for (const targetId of edge.targetIds) {
        sourceNeighbors.add(targetId);
        neighbors.get(targetId)?.add(sourceId);
      }
    }
  }

  let components = 0;
  const visited = new Set<string>();
  for (const nodeId of nodeIds) {
    if (visited.has(nodeId)) {
      continue;
    }

    components++;
    const frontier = [nodeId];
    while (frontier.length > 0) {
      const current = frontier.pop();
      if (current === undefined || visited.has(current)) {
        continue;
      }

      visited.add(current);
      for (const neighbor of neighbors.get(current) ?? []) {
        if (!visited.has(neighbor)) {
          frontier.push(neighbor);
        }
      }
    }
  }

  return Math.max(1, components);
}

function structuralBeta1(
  nodeIds: readonly string[],
  edges: readonly {
    sourceIds: readonly string[];
    targetIds: readonly string[];
  }[]
): number {
  const expandedEdgeCount = edges.reduce(
    (sum, edge) => sum + edge.sourceIds.length * edge.targetIds.length,
    0
  );

  return Math.max(
    0,
    expandedEdgeCount -
      nodeIds.length +
      countConnectedComponents(nodeIds, edges)
  );
}

function target(sample: Omit<Sample, 'target'>): number {
  return Math.abs(sample.leftInput) + Math.abs(sample.rightInput);
}

function buildSamples(axis: readonly number[]): Sample[] {
  return axis.flatMap((leftInput) =>
    axis.map((rightInput) => ({
      leftInput,
      rightInput,
      target: target({ leftInput, rightInput }),
    }))
  );
}

function initializeBlockParameters(seed: number): BlockParameters[] {
  const random = createDeterministicRandom(seed);
  return blockSpecs.map(() => ({
    outerGateWeight: 0.8 + random() * 0.4,
    outerGateBias: (random() - 0.5) * 0.2,
    heads: headSpecs.map(() => ({
      gateWeight: 0.8 + random() * 0.6,
      gateBias: (random() - 0.5) * 0.2,
      valueWeight: 0.3 + random() * 0.9,
      valueBias: (random() - 0.5) * 0.1,
    })),
  }));
}

function selectMoaBlockIndices(
  sample: Sample,
  activeBlockCount: number
): readonly number[] {
  const selected = [
    sample.leftInput >= 0 ? 0 : 1,
    sample.rightInput >= 0 ? 2 : 3,
  ];

  if (activeBlockCount <= selected.length) {
    return selected.slice(0, activeBlockCount);
  }

  const remaining = [0, 1, 2, 3].filter((index) => !selected.includes(index));
  remaining.sort((left, right) => {
    const leftScore = Math.abs(blockSpecs[left]?.signedProjection(sample) ?? 0);
    const rightScore = Math.abs(
      blockSpecs[right]?.signedProjection(sample) ?? 0
    );
    return rightScore - leftScore;
  });

  return [
    ...selected,
    ...remaining.slice(0, activeBlockCount - selected.length),
  ];
}

function selectMoaHeadIndices(
  axisMagnitude: number,
  activeHeadCount: number
): readonly number[] {
  return headSpecs
    .map((head, index) => ({
      index,
      distance: Math.abs(axisMagnitude - head.magnitudeCenter),
    }))
    .sort((left, right) => left.distance - right.distance)
    .slice(0, activeHeadCount)
    .map((entry) => entry.index);
}

function activateBlocks(
  family: MoaTransformerFamily,
  parameters: readonly BlockParameters[],
  sample: Sample,
  config: Pick<
    MoaTransformerShootoutConfig,
    'activeBlockCount' | 'activeHeadCount'
  >
): readonly BlockActivation[] {
  const activeBlockIndices =
    family === 'regular'
      ? blockSpecs.map((_, index) => index)
      : selectMoaBlockIndices(sample, config.activeBlockCount);

  return blockSpecs.map((block, blockIndex) => {
    const blockParameters = parameters[blockIndex];
    if (!blockParameters) {
      throw new Error(`Missing block parameters for index ${blockIndex}`);
    }

    const active = activeBlockIndices.includes(blockIndex);
    const signedInput = block.signedProjection(sample);
    const axisInput = block.axisProjection(sample);
    const axisMagnitude = Math.abs(axisInput);
    if (!active) {
      return {
        outerGate: 0,
        blockOutput: 0,
        signedInput,
        axisInput,
        active: false,
        selectedHeadCount: 0,
        headActivations: headSpecs.map(() => createDormantHeadActivation()),
      };
    }

    const outerGate = sigmoid(
      blockParameters.outerGateWeight * signedInput +
        blockParameters.outerGateBias
    );
    const activeHeadIndices =
      family === 'regular'
        ? headSpecs.map((_, index) => index)
        : selectMoaHeadIndices(axisMagnitude, config.activeHeadCount);
    const activeHeadSet = new Set(activeHeadIndices);

    const headActivations = headSpecs.map((head, headIndex) => {
      if (!activeHeadSet.has(headIndex)) {
        return createDormantHeadActivation();
      }

      const headParameters = blockParameters.heads[headIndex];
      if (!headParameters) {
        throw new Error(
          `Missing head parameters for block ${block.id} head ${head.id}`
        );
      }

      const headInput =
        signedInput * head.inputScale -
        Math.abs(axisMagnitude - head.magnitudeCenter) * 1.5;
      const valueInput = axisMagnitude * head.valueScale;
      const gate = sigmoid(
        headParameters.gateWeight * headInput + headParameters.gateBias
      );
      const value =
        headParameters.valueWeight * valueInput + headParameters.valueBias;
      return {
        gate,
        value,
        output: gate * value,
        headInput,
        valueInput,
        selected: activeHeadSet.has(headIndex),
      };
    });

    const selectedHeadOutputs = headActivations.reduce(
      (sum, activation) =>
        activation.selected ? sum + activation.output : sum,
      0
    );

    return {
      outerGate,
      blockOutput: outerGate * selectedHeadOutputs,
      signedInput,
      axisInput,
      active,
      selectedHeadCount: headActivations.filter(
        (activation) => activation.selected
      ).length,
      headActivations,
    };
  });
}

function predict(
  family: MoaTransformerFamily,
  parameters: readonly BlockParameters[],
  sample: Sample,
  config: Pick<
    MoaTransformerShootoutConfig,
    'activeBlockCount' | 'activeHeadCount'
  >
): PredictionResult {
  const blockActivations = activateBlocks(family, parameters, sample, config);
  return {
    prediction: blockActivations.reduce(
      (sum, activation) => sum + activation.blockOutput,
      0
    ),
    activeBlockCount: blockActivations.filter((activation) => activation.active)
      .length,
    activeHeadCount: blockActivations.reduce(
      (sum, activation) => sum + activation.selectedHeadCount,
      0
    ),
    blockActivations,
  };
}

function updateParameters(
  family: MoaTransformerFamily,
  parameters: BlockParameters[],
  sample: Sample,
  config: Pick<
    MoaTransformerShootoutConfig,
    'activeBlockCount' | 'activeHeadCount' | 'learningRate'
  >
): void {
  const result = predict(family, parameters, sample, config);
  const error = result.prediction - sample.target;

  for (const [blockIndex, activation] of result.blockActivations.entries()) {
    if (!activation?.active) {
      continue;
    }

    const blockParameters = parameters[blockIndex];
    if (!blockParameters) {
      continue;
    }

    const selectedHeadOutputs = activation.headActivations
      .filter((head) => head.selected)
      .reduce((sum, head) => sum + head.output, 0);
    const outerPrime = activation.outerGate * (1 - activation.outerGate);

    blockParameters.outerGateWeight -=
      config.learningRate *
      error *
      outerPrime *
      activation.signedInput *
      selectedHeadOutputs;
    blockParameters.outerGateBias -=
      config.learningRate * error * outerPrime * selectedHeadOutputs;

    for (const [
      headIndex,
      headActivation,
    ] of activation.headActivations.entries()) {
      if (!headActivation?.selected) {
        continue;
      }

      const headParameters = blockParameters.heads[headIndex];
      if (!headParameters) {
        continue;
      }

      const gatePrime = headActivation.gate * (1 - headActivation.gate);
      headParameters.valueWeight -=
        config.learningRate *
        error *
        activation.outerGate *
        headActivation.gate *
        headActivation.valueInput;
      headParameters.valueBias -=
        config.learningRate *
        error *
        activation.outerGate *
        headActivation.gate;
      headParameters.gateWeight -=
        config.learningRate *
        error *
        activation.outerGate *
        gatePrime *
        headActivation.headInput *
        headActivation.value;
      headParameters.gateBias -=
        config.learningRate *
        error *
        activation.outerGate *
        gatePrime *
        headActivation.value;
    }
  }
}

function evaluateSamplesDirect(
  family: MoaTransformerFamily,
  parameters: readonly BlockParameters[],
  samples: readonly Sample[],
  config: Pick<
    MoaTransformerShootoutConfig,
    | 'activeBlockCount'
    | 'activeHeadCount'
    | 'dualActiveThreshold'
    | 'exactPredictionTolerance'
  >
): EvaluationMetrics {
  if (samples.length === 0) {
    return {
      meanSquaredError: 0,
      exactWithinToleranceFraction: 0,
      dualActiveRegionMeanAbsoluteError: 0,
      activeBlockCount: 0,
      activeHeadCount: 0,
    };
  }

  let squaredErrorSum = 0;
  let exactCount = 0;
  let dualActiveAbsoluteErrorSum = 0;
  let dualActiveCount = 0;
  let activeBlockCountSum = 0;
  let activeHeadCountSum = 0;

  for (const sample of samples) {
    const prediction = predict(family, parameters, sample, config);
    const error = prediction.prediction - sample.target;
    squaredErrorSum += error * error;
    activeBlockCountSum += prediction.activeBlockCount;
    activeHeadCountSum += prediction.activeHeadCount;

    if (Math.abs(error) <= config.exactPredictionTolerance) {
      exactCount++;
    }

    if (
      Math.abs(sample.leftInput) >= config.dualActiveThreshold &&
      Math.abs(sample.rightInput) >= config.dualActiveThreshold
    ) {
      dualActiveAbsoluteErrorSum += Math.abs(error);
      dualActiveCount++;
    }
  }

  return {
    meanSquaredError: squaredErrorSum / samples.length,
    exactWithinToleranceFraction: exactCount / samples.length,
    dualActiveRegionMeanAbsoluteError:
      dualActiveCount === 0 ? 0 : dualActiveAbsoluteErrorSum / dualActiveCount,
    activeBlockCount: activeBlockCountSum / samples.length,
    activeHeadCount: activeHeadCountSum / samples.length,
  };
}

function encodePayload(values: readonly number[]): Uint8Array {
  const floatValues = new Float64Array(values.length);
  values.forEach((value, index) => {
    floatValues[index] = value;
  });
  return new Uint8Array(floatValues.buffer.slice(0));
}

function decodePayload(payload: Uint8Array): readonly number[] {
  const aligned = payload.slice();
  return Array.from(
    new Float64Array(aligned.buffer, aligned.byteOffset, aligned.byteLength / 8)
  );
}

function framesEqual(left: FlowFrame, right: FlowFrame): boolean {
  if (
    left.streamId !== right.streamId ||
    left.sequence !== right.sequence ||
    left.flags !== right.flags ||
    left.payload.length !== right.payload.length
  ) {
    return false;
  }

  for (let index = 0; index < left.payload.length; index++) {
    if (left.payload[index] !== right.payload[index]) {
      return false;
    }
  }

  return true;
}

function buildFlowFrames(
  sampleIndex: number,
  prediction: PredictionResult
): readonly FlowFrame[] {
  const streamOffset = sampleIndex * 128;

  return prediction.blockActivations.flatMap((blockActivation, blockIndex) => {
    if (!blockActivation.active) {
      return [];
    }

    const block = blockSpecs[blockIndex];
    if (!block) {
      throw new Error(`Missing block spec for index ${blockIndex}`);
    }

    return blockActivation.headActivations.flatMap(
      (headActivation, headIndex) => {
        if (!headActivation.selected) {
          return [];
        }

        const streamId = streamOffset + block.streamBaseId * 10 + headIndex;
        return [
          {
            streamId,
            sequence: 0,
            flags: FORK,
            payload: encodePayload([
              blockActivation.signedInput,
              Math.abs(blockActivation.axisInput),
              blockActivation.outerGate,
            ]),
          },
          {
            streamId,
            sequence: 1,
            flags: RACE,
            payload: encodePayload([
              headActivation.headInput,
              headActivation.gate,
            ]),
          },
          {
            streamId,
            sequence: 2,
            flags: RACE,
            payload: encodePayload([
              headActivation.valueInput,
              headActivation.value,
            ]),
          },
          {
            streamId,
            sequence: 3,
            flags: FOLD | FIN,
            payload: encodePayload([
              blockActivation.outerGate * headActivation.output,
            ]),
          },
        ] satisfies readonly FlowFrame[];
      }
    );
  });
}

function frameSort(left: FlowFrame, right: FlowFrame): number {
  return left.streamId - right.streamId || left.sequence - right.sequence;
}

function reconstructPrediction(frames: readonly FlowFrame[]): number {
  return frames
    .filter((frame) => frame.sequence === 3)
    .reduce((sum, frame) => {
      const [output] = decodePayload(frame.payload);
      return sum + (output ?? 0);
    }, 0);
}

function predictionsMatchWithinTolerance(
  left: number,
  right: number,
  tolerance = 1e-9
): boolean {
  return Math.abs(left - right) <= tolerance;
}

function framedPredict(
  family: MoaTransformerFamily,
  parameters: readonly BlockParameters[],
  sample: Sample,
  sampleIndex: number,
  config: Pick<
    MoaTransformerShootoutConfig,
    'activeBlockCount' | 'activeHeadCount'
  >
): FramedPredictionResult {
  const direct = predict(family, parameters, sample, config);
  const originalFrames = buildFlowFrames(sampleIndex, direct);
  const codec = FlowCodec.createSync();

  let codecExactCount = 0;
  const decodedFrames = originalFrames.map((frame) => {
    const encoded = codec.encode(frame);
    const decoded = codec.decode(encoded).frame;
    if (framesEqual(frame, decoded)) {
      codecExactCount++;
    }
    return decoded;
  });

  const reassembler = new FrameReassembler();
  const deliveredFrames: FlowFrame[] = [];
  for (const frame of [...decodedFrames].reverse()) {
    deliveredFrames.push(...reassembler.push(frame));
  }

  const sortedOriginal = [...originalFrames].sort(frameSort);
  const sortedDelivered = [...deliveredFrames].sort(frameSort);
  let reassemblyExactCount = 0;
  for (let index = 0; index < sortedOriginal.length; index++) {
    const original = sortedOriginal[index];
    const delivered = sortedDelivered[index];
    if (original && delivered && framesEqual(original, delivered)) {
      reassemblyExactCount++;
    }
  }

  return {
    prediction: direct.prediction,
    codecRoundTripExactFraction:
      originalFrames.length === 0 ? 0 : codecExactCount / originalFrames.length,
    reassemblyExactFraction:
      originalFrames.length === 0
        ? 0
        : reassemblyExactCount / originalFrames.length,
    foldInvarianceExactFraction: predictionsMatchWithinTolerance(
      reconstructPrediction(sortedDelivered),
      direct.prediction
    )
      ? 1
      : 0,
    activeBlockCount: direct.activeBlockCount,
    activeHeadCount: direct.activeHeadCount,
    frameCount: originalFrames.length,
    payloadBytes: originalFrames.reduce(
      (sum, frame) => sum + frame.payload.length,
      0
    ),
  };
}

function evaluateSamplesWithTransport(
  family: MoaTransformerFamily,
  parameters: readonly BlockParameters[],
  samples: readonly Sample[],
  config: Pick<
    MoaTransformerShootoutConfig,
    | 'activeBlockCount'
    | 'activeHeadCount'
    | 'dualActiveThreshold'
    | 'exactPredictionTolerance'
  >
): EvaluationMetrics & TransportMetrics {
  if (samples.length === 0) {
    return {
      meanSquaredError: 0,
      exactWithinToleranceFraction: 0,
      dualActiveRegionMeanAbsoluteError: 0,
      activeBlockCount: 0,
      activeHeadCount: 0,
      codecRoundTripExactFraction: 0,
      reassemblyExactFraction: 0,
      foldInvarianceExactFraction: 0,
      frameCount: 0,
      payloadBytes: 0,
    };
  }

  let squaredErrorSum = 0;
  let exactCount = 0;
  let dualActiveAbsoluteErrorSum = 0;
  let dualActiveCount = 0;
  let activeBlockCountSum = 0;
  let activeHeadCountSum = 0;
  let codecRoundTripSum = 0;
  let reassemblySum = 0;
  let foldInvarianceSum = 0;
  let frameCountSum = 0;
  let payloadBytesSum = 0;

  for (const [sampleIndex, sample] of samples.entries()) {
    const framed = framedPredict(
      family,
      parameters,
      sample,
      sampleIndex,
      config
    );
    const error = framed.prediction - sample.target;
    squaredErrorSum += error * error;
    activeBlockCountSum += framed.activeBlockCount;
    activeHeadCountSum += framed.activeHeadCount;
    codecRoundTripSum += framed.codecRoundTripExactFraction;
    reassemblySum += framed.reassemblyExactFraction;
    foldInvarianceSum += framed.foldInvarianceExactFraction;
    frameCountSum += framed.frameCount;
    payloadBytesSum += framed.payloadBytes;

    if (Math.abs(error) <= config.exactPredictionTolerance) {
      exactCount++;
    }

    if (
      Math.abs(sample.leftInput) >= config.dualActiveThreshold &&
      Math.abs(sample.rightInput) >= config.dualActiveThreshold
    ) {
      dualActiveAbsoluteErrorSum += Math.abs(error);
      dualActiveCount++;
    }
  }

  return {
    meanSquaredError: squaredErrorSum / samples.length,
    exactWithinToleranceFraction: exactCount / samples.length,
    dualActiveRegionMeanAbsoluteError:
      dualActiveCount === 0 ? 0 : dualActiveAbsoluteErrorSum / dualActiveCount,
    activeBlockCount: activeBlockCountSum / samples.length,
    activeHeadCount: activeHeadCountSum / samples.length,
    codecRoundTripExactFraction: codecRoundTripSum / samples.length,
    reassemblyExactFraction: reassemblySum / samples.length,
    foldInvarianceExactFraction: foldInvarianceSum / samples.length,
    frameCount: frameCountSum / samples.length,
    payloadBytes: payloadBytesSum / samples.length,
  };
}

function summarizeSeedMetrics(
  metrics: readonly MoaTransformerSeedMetrics[]
): Omit<
  MoaTransformerFamilyReport,
  | 'topologyPath'
  | 'topologyNodeCount'
  | 'topologyEdgeCount'
  | 'structuralBeta1'
  | 'transformerletCount'
  | 'headChainCount'
  | 'rotationStageCount'
  | 'parameterCount'
  | 'usesOuterRouter'
  | 'samplePredictions'
> {
  const evalMeanSquaredErrors = metrics.map(
    (entry) => entry.evalMeanSquaredError
  );
  const exactFractions = metrics.map(
    (entry) => entry.exactWithinToleranceFraction
  );
  const dualActiveErrors = metrics.map(
    (entry) => entry.dualActiveRegionMeanAbsoluteError
  );
  const activeHeadCounts = metrics.map((entry) => entry.activeHeadCount);

  return {
    seedMetrics: metrics,
    meanTrainMeanSquaredError: mean(
      metrics.map((entry) => entry.trainMeanSquaredError)
    ),
    stdevTrainMeanSquaredError: stdev(
      metrics.map((entry) => entry.trainMeanSquaredError)
    ),
    meanEvalMeanSquaredError: mean(evalMeanSquaredErrors),
    evalMeanSquaredErrorCi95: bootstrapMeanConfidenceInterval(
      evalMeanSquaredErrors,
      { seed: 0x4d4f4152 }
    ),
    stdevEvalMeanSquaredError: stdev(evalMeanSquaredErrors),
    meanExactWithinToleranceFraction: mean(exactFractions),
    exactWithinToleranceFractionCi95: bootstrapMeanConfidenceInterval(
      exactFractions,
      { seed: 0x5452414e }
    ),
    stdevExactWithinToleranceFraction: stdev(exactFractions),
    meanDualActiveRegionMeanAbsoluteError: mean(dualActiveErrors),
    dualActiveRegionMeanAbsoluteErrorCi95: bootstrapMeanConfidenceInterval(
      dualActiveErrors,
      { seed: 0x53484f54 }
    ),
    stdevDualActiveRegionMeanAbsoluteError: stdev(dualActiveErrors),
    meanTrainWallTimeMs: mean(metrics.map((entry) => entry.trainWallTimeMs)),
    meanEvalWallTimeMs: mean(metrics.map((entry) => entry.evalWallTimeMs)),
    meanEvalSamplesPerSecond: mean(
      metrics.map((entry) => entry.evalSamplesPerSecond)
    ),
    meanCodecRoundTripExactFraction: mean(
      metrics.map((entry) => entry.codecRoundTripExactFraction)
    ),
    meanReassemblyExactFraction: mean(
      metrics.map((entry) => entry.reassemblyExactFraction)
    ),
    meanFoldInvarianceExactFraction: mean(
      metrics.map((entry) => entry.foldInvarianceExactFraction)
    ),
    meanActiveBlockCount: mean(metrics.map((entry) => entry.activeBlockCount)),
    meanActiveHeadCount: mean(activeHeadCounts),
    meanHeadUtilizationFraction:
      mean(activeHeadCounts) / (blockSpecs.length * headSpecs.length),
    meanFrameCount: mean(metrics.map((entry) => entry.frameCount)),
    meanPayloadBytes: mean(metrics.map((entry) => entry.payloadBytes)),
    computeAdjustedExactFraction:
      mean(activeHeadCounts) === 0
        ? 0
        : mean(exactFractions) / mean(activeHeadCounts),
    computeAdjustedEvalLoss:
      mean(evalMeanSquaredErrors) * Math.max(1, mean(activeHeadCounts)),
  };
}

function samplePredictions(
  family: MoaTransformerFamily,
  trainedParameterSets: readonly (readonly BlockParameters[])[],
  samplePoints: readonly (readonly [number, number])[],
  config: Pick<
    MoaTransformerShootoutConfig,
    'activeBlockCount' | 'activeHeadCount'
  >
): MoaTransformerSamplePrediction[] {
  return samplePoints.map(([leftInput, rightInput]) => {
    const sample: Sample = {
      leftInput,
      rightInput,
      target: target({ leftInput, rightInput }),
    };
    const predictions = trainedParameterSets.map((parameterSet, seedIndex) =>
      framedPredict(family, parameterSet, sample, seedIndex, config)
    );

    return {
      leftInput,
      rightInput,
      target: sample.target,
      meanPrediction: mean(predictions.map((entry) => entry.prediction)),
      predictionStddev: stdev(predictions.map((entry) => entry.prediction)),
      meanActiveBlockCount: mean(
        predictions.map((entry) => entry.activeBlockCount)
      ),
      meanActiveHeadCount: mean(
        predictions.map((entry) => entry.activeHeadCount)
      ),
      meanFrameCount: mean(predictions.map((entry) => entry.frameCount)),
    };
  });
}

/**
 * Handles the gnosis make Default Moa Transformer Shootout Config workflow.
 */
export function makeDefaultMoaTransformerShootoutConfig(): MoaTransformerShootoutConfig {
  return {
    label: 'gnosis-moa-transformer-shootout-v1',
    seedCount: 12,
    epochs: 1200,
    learningRate: 0.018,
    exactPredictionTolerance: 0.08,
    dualActiveThreshold: 0.4,
    frameStageCount: 4,
    activeBlockCount: 2,
    activeHeadCount: 2,
    trainAxis: [
      -1.5, -1.25, -1.0, -0.75, -0.5, -0.25, 0, 0.25, 0.5, 0.75, 1.0, 1.25, 1.5,
    ],
    evalAxis: [-1.4, -1.1, -0.8, -0.4, -0.1, 0.1, 0.4, 0.8, 1.1, 1.4],
    samplePoints: [
      [-1.2, -1.2],
      [-1.2, 1.2],
      [1.2, -1.2],
      [1.2, 1.2],
      [1.2, 0],
    ],
  };
}

async function loadTopology(
  family: MoaTransformerFamily
): Promise<ParsedTopology> {
  const path = DEFAULT_MOA_TRANSFORMER_TOPOLOGY_FILES[family];
  const source = await readFile(path, 'utf8');
  const program = parseGgProgram(lowerUfcsSource(source));
  const outerFoldEdges = program.edges.filter(
    (edge) =>
      edge.type.toUpperCase() === 'FOLD' &&
      edge.properties.boundary === 'outer-whip'
  );
  if (outerFoldEdges.length !== 1) {
    throw new Error(
      `Expected exactly one outer-whip FOLD edge in ${path}, found ${outerFoldEdges.length}`
    );
  }

  const transformerletNodes = program.nodes.filter((node) =>
    node.labels.some((label) => label.includes('Transformerlet'))
  );
  if (transformerletNodes.length !== 4) {
    throw new Error(`Expected exactly four Transformerlet nodes in ${path}`);
  }

  const headChainNodes = program.nodes.filter((node) =>
    node.labels.includes('AttentionHeadChain')
  );
  if (headChainNodes.length !== 16) {
    throw new Error(
      `Expected exactly sixteen AttentionHeadChain nodes in ${path}`
    );
  }

  const rotationNodes = program.nodes.filter((node) =>
    node.labels.includes('RotationScheduler')
  );
  if (rotationNodes.length < 5) {
    throw new Error(
      `Expected at least five RotationScheduler nodes in ${path}`
    );
  }

  const parameterCount = transformerletNodes.reduce((sum, node) => {
    const rawParameters = Number(node.properties.parameters ?? 0);
    return sum + (Number.isFinite(rawParameters) ? rawParameters : 0);
  }, 0);

  return {
    family,
    path,
    metrics: {
      nodeCount: program.nodes.length,
      edgeCount: program.edges.length,
      structuralBeta1: structuralBeta1(
        program.nodes.map((node) => node.id),
        program.edges.map((edge) => ({
          sourceIds: edge.sourceIds,
          targetIds: edge.targetIds,
        }))
      ),
      transformerletCount: transformerletNodes.length,
      headChainCount: headChainNodes.length,
      rotationStageCount: rotationNodes.reduce((maxStages, node) => {
        const rawStages = Number(node.properties.stages ?? 0);
        return Number.isFinite(rawStages)
          ? Math.max(maxStages, rawStages)
          : maxStages;
      }, 0),
      parameterCount,
      usesOuterRouter: program.nodes.some((node) =>
        node.labels.includes('RoutingGate')
      ),
    },
  };
}

async function loadDefaultTopologySuite(): Promise<
  Record<MoaTransformerFamily, ParsedTopology>
> {
  const entries = await Promise.all(
    (
      Object.keys(
        DEFAULT_MOA_TRANSFORMER_TOPOLOGY_FILES
      ) as MoaTransformerFamily[]
    ).map(async (family) => [family, await loadTopology(family)] as const)
  );

  return Object.fromEntries(entries) as Record<
    MoaTransformerFamily,
    ParsedTopology
  >;
}

async function loadDefaultTopologySuiteCached(): Promise<
  Record<MoaTransformerFamily, ParsedTopology>
> {
  defaultTopologySuitePromise ??= loadDefaultTopologySuite();
  return defaultTopologySuitePromise;
}

function assertComparableCapacity(
  topologies: Record<MoaTransformerFamily, ParsedTopology>
): GnosisMoaTransformerShootoutReport['sharedCapacity'] {
  const regular = topologies.regular.metrics;
  const moa = topologies.moa.metrics;
  if (
    regular.transformerletCount !== moa.transformerletCount ||
    regular.headChainCount !== moa.headChainCount ||
    regular.rotationStageCount !== moa.rotationStageCount ||
    regular.parameterCount !== moa.parameterCount
  ) {
    throw new Error('Regular and MoA topologies are not capacity-matched');
  }

  return {
    transformerletCount: regular.transformerletCount,
    headChainCount: regular.headChainCount,
    rotationStageCount: regular.rotationStageCount,
    parameterCount: regular.parameterCount,
  };
}

function trainFamily(
  family: MoaTransformerFamily,
  config: MoaTransformerShootoutConfig,
  trainSamples: readonly Sample[],
  evalSamples: readonly Sample[]
): {
  readonly seedMetrics: readonly MoaTransformerSeedMetrics[];
  readonly trainedParameters: readonly (readonly BlockParameters[])[];
} {
  const seedMetrics: MoaTransformerSeedMetrics[] = [];
  const trainedParameters: BlockParameters[][] = [];

  for (let seedIndex = 0; seedIndex < config.seedCount; seedIndex++) {
    const seed = seedIndex + 1;
    const parameters = initializeBlockParameters(seed);
    const random = createDeterministicRandom((seed * 0x9e3779b1) >>> 0);

    const trainStart = performance.now();
    for (let epoch = 0; epoch < config.epochs; epoch++) {
      const parityOffset = Math.floor(random() * trainSamples.length);
      for (let offset = 0; offset < trainSamples.length; offset++) {
        const sample =
          trainSamples[(offset + parityOffset) % trainSamples.length];
        if (sample) {
          updateParameters(family, parameters, sample, config);
        }
      }
    }
    const trainWallTimeMs = performance.now() - trainStart;

    const trainEvaluation = evaluateSamplesDirect(
      family,
      parameters,
      trainSamples,
      config
    );
    const evalStart = performance.now();
    const evalEvaluation = evaluateSamplesWithTransport(
      family,
      parameters,
      evalSamples,
      config
    );
    const evalWallTimeMs = performance.now() - evalStart;

    seedMetrics.push({
      seed,
      trainMeanSquaredError: trainEvaluation.meanSquaredError,
      evalMeanSquaredError: evalEvaluation.meanSquaredError,
      exactWithinToleranceFraction: evalEvaluation.exactWithinToleranceFraction,
      dualActiveRegionMeanAbsoluteError:
        evalEvaluation.dualActiveRegionMeanAbsoluteError,
      trainWallTimeMs,
      evalWallTimeMs,
      evalSamplesPerSecond:
        evalWallTimeMs <= 0 ? 0 : evalSamples.length / (evalWallTimeMs / 1000),
      codecRoundTripExactFraction: evalEvaluation.codecRoundTripExactFraction,
      reassemblyExactFraction: evalEvaluation.reassemblyExactFraction,
      foldInvarianceExactFraction: evalEvaluation.foldInvarianceExactFraction,
      activeBlockCount: evalEvaluation.activeBlockCount,
      activeHeadCount: evalEvaluation.activeHeadCount,
      frameCount: evalEvaluation.frameCount,
      payloadBytes: evalEvaluation.payloadBytes,
    });
    trainedParameters.push(
      parameters.map((block) => ({
        outerGateWeight: block.outerGateWeight,
        outerGateBias: block.outerGateBias,
        heads: block.heads.map((head) => ({ ...head })),
      }))
    );
  }

  return {
    seedMetrics,
    trainedParameters,
  };
}

function buildFamilyReport(
  family: MoaTransformerFamily,
  topologies: Record<MoaTransformerFamily, ParsedTopology>,
  config: MoaTransformerShootoutConfig,
  trainSamples: readonly Sample[],
  evalSamples: readonly Sample[]
): MoaTransformerFamilyReport {
  const trained = trainFamily(family, config, trainSamples, evalSamples);
  const summary = summarizeSeedMetrics(trained.seedMetrics);
  const metrics = topologies[family].metrics;

  return {
    topologyPath: topologies[family].path,
    topologyNodeCount: metrics.nodeCount,
    topologyEdgeCount: metrics.edgeCount,
    structuralBeta1: metrics.structuralBeta1,
    transformerletCount: metrics.transformerletCount,
    headChainCount: metrics.headChainCount,
    rotationStageCount: metrics.rotationStageCount,
    parameterCount: metrics.parameterCount,
    usesOuterRouter: metrics.usesOuterRouter,
    ...summary,
    samplePredictions: samplePredictions(
      family,
      trained.trainedParameters,
      config.samplePoints,
      config
    ),
  };
}

/**
 * Handles the gnosis run Gnosis Moa Transformer Family Benchmark workflow.
 */
export async function runGnosisMoaTransformerFamilyBenchmark(
  family: MoaTransformerFamily,
  config: MoaTransformerShootoutConfig = makeDefaultMoaTransformerShootoutConfig()
): Promise<MoaTransformerFamilyReport> {
  const topologies = await loadDefaultTopologySuiteCached();
  assertComparableCapacity(topologies);
  const trainSamples = buildSamples(config.trainAxis);
  const evalSamples = buildSamples(config.evalAxis);
  return buildFamilyReport(
    family,
    topologies,
    config,
    trainSamples,
    evalSamples
  );
}

/**
 * Handles the gnosis run Gnosis Moa Transformer Shootout Benchmark workflow.
 */
export async function runGnosisMoaTransformerShootoutBenchmark(
  config: MoaTransformerShootoutConfig = makeDefaultMoaTransformerShootoutConfig()
): Promise<GnosisMoaTransformerShootoutReport> {
  const topologies = await loadDefaultTopologySuiteCached();
  const sharedCapacity = assertComparableCapacity(topologies);
  const trainSamples = buildSamples(config.trainAxis);
  const evalSamples = buildSamples(config.evalAxis);

  const families = {} as Record<
    MoaTransformerFamily,
    MoaTransformerFamilyReport
  >;
  for (const family of Object.keys(topologies) as MoaTransformerFamily[]) {
    families[family] = buildFamilyReport(
      family,
      topologies,
      config,
      trainSamples,
      evalSamples
    );
  }

  return {
    label: config.label,
    config: {
      seedCount: config.seedCount,
      epochs: config.epochs,
      learningRate: config.learningRate,
      exactPredictionTolerance: config.exactPredictionTolerance,
      dualActiveThreshold: config.dualActiveThreshold,
      frameStageCount: config.frameStageCount,
      activeBlockCount: config.activeBlockCount,
      activeHeadCount: config.activeHeadCount,
      trainSampleCount: trainSamples.length,
      evalSampleCount: evalSamples.length,
      targetFamily: 'quadrant-l1-rotated-transformer',
    },
    families,
    sharedCapacity,
    winnerByEvalMeanSquaredError:
      families.moa.meanEvalMeanSquaredError <=
      families.regular.meanEvalMeanSquaredError
        ? 'moa'
        : 'regular',
    winnerByComputeAdjustedExactFraction:
      families.moa.computeAdjustedExactFraction >=
      families.regular.computeAdjustedExactFraction
        ? 'moa'
        : 'regular',
    moaUsesLessCompute:
      families.moa.meanActiveHeadCount < families.regular.meanActiveHeadCount &&
      families.moa.meanFrameCount < families.regular.meanFrameCount,
    rawAccuracyCompetitive:
      families.moa.meanEvalMeanSquaredError <=
        families.regular.meanEvalMeanSquaredError + 0.03 &&
      families.moa.meanExactWithinToleranceFraction >=
        families.regular.meanExactWithinToleranceFraction - 0.05,
  };
}

function formatSamplePair(sample: MoaTransformerSamplePrediction): string {
  return `(${sample.leftInput.toFixed(2)}, ${sample.rightInput.toFixed(2)})`;
}

function formatInterval(interval: ConfidenceInterval): string {
  return `[${interval.low.toFixed(3)}, ${interval.high.toFixed(3)}]`;
}

/**
 * Handles the gnosis render Gnosis Moa Transformer Shootout Markdown workflow.
 */
export function renderGnosisMoaTransformerShootoutMarkdown(
  report: GnosisMoaTransformerShootoutReport
): string {
  const lines: string[] = [];
  lines.push('# Gnosis MoA Transformer Shootout');
  lines.push('');
  lines.push(`- Label: \`${report.label}\``);
  lines.push(`- Target family: \`${report.config.targetFamily}\``);
  lines.push(`- Seeds: \`${report.config.seedCount}\``);
  lines.push(`- Epochs: \`${report.config.epochs}\``);
  lines.push(`- Learning rate: \`${report.config.learningRate}\``);
  lines.push(`- Active MoA blocks: \`${report.config.activeBlockCount}\``);
  lines.push(
    `- Active MoA heads per live block: \`${report.config.activeHeadCount}\``
  );
  lines.push(
    `- Shared parameter count: \`${report.sharedCapacity.parameterCount}\``
  );
  lines.push(
    `- Shared transformerlets: \`${report.sharedCapacity.transformerletCount}\``
  );
  lines.push(
    `- Shared head chains: \`${report.sharedCapacity.headChainCount}\``
  );
  lines.push(`- Eval-MSE winner: \`${report.winnerByEvalMeanSquaredError}\``);
  lines.push(
    `- Compute-adjusted exact winner: \`${report.winnerByComputeAdjustedExactFraction}\``
  );
  lines.push(
    `- MoA uses less compute: \`${report.moaUsesLessCompute ? 'yes' : 'no'}\``
  );
  lines.push(
    `- MoA remains accuracy-competitive: \`${
      report.rawAccuracyCompetitive ? 'yes' : 'no'
    }\``
  );
  lines.push('');
  lines.push('## Aggregated Metrics');
  lines.push('');
  lines.push(
    '| Family | Mean eval MSE | Eval MSE 95% CI | Exact-within-tolerance | Exact 95% CI | Dual-active abs error | Mean active blocks | Mean active heads | Mean frames/sample | Compute-adjusted exact | Compute-adjusted eval loss | Codec round-trip | Reassembly | Fold invariance |'
  );
  lines.push(
    '|---|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|'
  );

  for (const family of [
    'moa',
    'regular',
  ] satisfies readonly MoaTransformerFamily[]) {
    const metrics = report.families[family];
    lines.push(
      `| \`${family}\` | ${metrics.meanEvalMeanSquaredError.toFixed(
        3
      )} | ${formatInterval(
        metrics.evalMeanSquaredErrorCi95
      )} | ${metrics.meanExactWithinToleranceFraction.toFixed(
        3
      )} | ${formatInterval(
        metrics.exactWithinToleranceFractionCi95
      )} | ${metrics.meanDualActiveRegionMeanAbsoluteError.toFixed(
        3
      )} | ${metrics.meanActiveBlockCount.toFixed(
        1
      )} | ${metrics.meanActiveHeadCount.toFixed(
        1
      )} | ${metrics.meanFrameCount.toFixed(
        1
      )} | ${metrics.computeAdjustedExactFraction.toFixed(
        4
      )} | ${metrics.computeAdjustedEvalLoss.toFixed(
        3
      )} | ${metrics.meanCodecRoundTripExactFraction.toFixed(
        3
      )} | ${metrics.meanReassemblyExactFraction.toFixed(
        3
      )} | ${metrics.meanFoldInvarianceExactFraction.toFixed(3)} |`
    );
  }

  lines.push('');
  lines.push('## Timing');
  lines.push('');
  lines.push(
    '| Family | Mean train wall ms | Mean eval wall ms | Mean eval samples/sec |'
  );
  lines.push('|---|---:|---:|---:|');
  for (const family of [
    'moa',
    'regular',
  ] satisfies readonly MoaTransformerFamily[]) {
    const metrics = report.families[family];
    lines.push(
      `| \`${family}\` | ${metrics.meanTrainWallTimeMs.toFixed(
        2
      )} | ${metrics.meanEvalWallTimeMs.toFixed(
        2
      )} | ${metrics.meanEvalSamplesPerSecond.toFixed(1)} |`
    );
  }

  lines.push('');
  lines.push('## Sample Predictions');
  lines.push('');
  lines.push(
    '| Family | Input pair | Target | Mean prediction | Prediction σ | Mean active blocks | Mean active heads | Mean frames |'
  );
  lines.push('|---|---|---:|---:|---:|---:|---:|---:|');
  for (const family of [
    'moa',
    'regular',
  ] satisfies readonly MoaTransformerFamily[]) {
    for (const sample of report.families[family].samplePredictions) {
      lines.push(
        `| \`${family}\` | ${formatSamplePair(
          sample
        )} | ${sample.target.toFixed(3)} | ${sample.meanPrediction.toFixed(
          3
        )} | ${sample.predictionStddev.toFixed(
          3
        )} | ${sample.meanActiveBlockCount.toFixed(
          1
        )} | ${sample.meanActiveHeadCount.toFixed(
          1
        )} | ${sample.meanFrameCount.toFixed(1)} |`
      );
    }
  }

  lines.push('');
  lines.push(
    'Interpretation: the regular family keeps every rotated transformerlet and every rotated head chain alive on every sample, while the MoA family uses the same total capacity as a structured yoyo: Wallington Rotation opens the staged frontier, a sparse outer route keeps only one sign-specialist block per axis, and a local Worthington-style whip keeps only the nearest two magnitude heads inside each live block before the final additive collapse. The benchmark reports whether that structured sparsity stays accuracy-competitive while materially reducing active chains and flow frames.'
  );
  lines.push('');

  return `${lines.join('\n')}\n`;
}
