// @a0n/gnosis/src/benchmarks/moe-routing-benchmark.ts
import { readFile } from 'node:fs/promises';
import { fileURLToPath } from 'node:url';

import { parseGgProgram } from '@a0n/aeon-logic';
import { lowerUfcsSource } from '../ufcs';

import {
  bootstrapMeanConfidenceInterval,
  createDeterministicRandom,
  mean,
  shuffleInPlace,
  stdev,
  type ConfidenceInterval,
} from './statistics';

export type MiniMoeRoutingStrategy =
  | 'linear'
  | 'winner-take-all'
  | 'early-stop';

export interface MiniMoeRoutingConfig {
  readonly label: 'gnosis-moe-routing-benchmark-v1';
  readonly seedCount: number;
  readonly epochs: number;
  readonly learningRate: number;
  readonly exactPredictionTolerance: number;
  readonly dualActiveThreshold: number;
  readonly trainAxis: readonly number[];
  readonly evalAxis: readonly number[];
  readonly samplePoints: readonly (readonly [number, number])[];
}

export interface MiniMoeRoutingSeedMetrics {
  readonly seed: number;
  readonly trainMeanSquaredError: number;
  readonly evalMeanSquaredError: number;
  readonly exactWithinToleranceFraction: number;
  readonly dualActiveRegionMeanAbsoluteError: number;
}

export interface MiniMoeRoutingSamplePrediction {
  readonly leftInput: number;
  readonly rightInput: number;
  readonly target: number;
  readonly meanPrediction: number;
  readonly predictionStddev: number;
}

export interface MiniMoeRoutingStrategyReport {
  readonly topologyPath: string;
  readonly topologyNodeCount: number;
  readonly topologyEdgeCount: number;
  readonly structuralBeta1: number;
  readonly expertCount: number;
  readonly parameterCount: number;
  readonly seedMetrics: readonly MiniMoeRoutingSeedMetrics[];
  readonly meanTrainMeanSquaredError: number;
  readonly stdevTrainMeanSquaredError: number;
  readonly meanEvalMeanSquaredError: number;
  readonly evalMeanSquaredErrorCi95: ConfidenceInterval;
  readonly stdevEvalMeanSquaredError: number;
  readonly meanExactWithinToleranceFraction: number;
  readonly exactWithinToleranceFractionCi95: ConfidenceInterval;
  readonly stdevExactWithinToleranceFraction: number;
  readonly meanDualActiveRegionMeanAbsoluteError: number;
  readonly dualActiveRegionMeanAbsoluteErrorCi95: ConfidenceInterval;
  readonly stdevDualActiveRegionMeanAbsoluteError: number;
  readonly samplePredictions: readonly MiniMoeRoutingSamplePrediction[];
}

export interface MiniMoeRoutingTopologyMetrics {
  readonly nodeCount: number;
  readonly edgeCount: number;
  readonly structuralBeta1: number;
  readonly expertCount: number;
  readonly parameterCount: number;
}

export interface GnosisMiniMoeRoutingBenchmarkReport {
  readonly label: MiniMoeRoutingConfig['label'];
  readonly config: {
    readonly seedCount: number;
    readonly epochs: number;
    readonly learningRate: number;
    readonly exactPredictionTolerance: number;
    readonly dualActiveThreshold: number;
    readonly trainSampleCount: number;
    readonly evalSampleCount: number;
    readonly targetFamily: 'quadrant-l1-routing';
  };
  readonly topology: MiniMoeRoutingTopologyMetrics;
  readonly strategies: Record<
    MiniMoeRoutingStrategy,
    MiniMoeRoutingStrategyReport
  >;
  readonly rankingByEvalMeanSquaredError: readonly MiniMoeRoutingStrategy[];
  readonly predictedRankingMatches: boolean;
}

interface MiniMoeRoutingSample {
  readonly leftInput: number;
  readonly rightInput: number;
  readonly target: number;
}

interface ExpertParameters {
  gateWeight: number;
  gateBias: number;
  valueWeight: number;
  valueBias: number;
}

interface ParsedTopology {
  readonly strategy: MiniMoeRoutingStrategy;
  readonly path: string;
  readonly metrics: MiniMoeRoutingTopologyMetrics;
}

interface ExpertSpec {
  readonly id: string;
  readonly signedProjection: (sample: MiniMoeRoutingSample) => number;
  readonly axisProjection: (sample: MiniMoeRoutingSample) => number;
}

interface ExpertActivation {
  readonly gate: number;
  readonly value: number;
  readonly output: number;
  readonly signedInput: number;
  readonly axisInput: number;
}

interface PredictionResult {
  readonly prediction: number;
  readonly selectedExpertIndex: number;
  readonly expertActivations: readonly ExpertActivation[];
}

const DEFAULT_MINI_MOE_ROUTING_TOPOLOGY_URLS: Record<
  MiniMoeRoutingStrategy,
  URL
> = {
  linear: new URL(
    '../../examples/benchmarks/moe-routing-linear.gg',
    import.meta.url
  ),
  'winner-take-all': new URL(
    '../../examples/benchmarks/moe-routing-winner-take-all.gg',
    import.meta.url
  ),
  'early-stop': new URL(
    '../../examples/benchmarks/moe-routing-early-stop.gg',
    import.meta.url
  ),
};

export const DEFAULT_MINI_MOE_ROUTING_TOPOLOGY_FILES = Object.fromEntries(
  (
    Object.entries(DEFAULT_MINI_MOE_ROUTING_TOPOLOGY_URLS) as Array<
      readonly [MiniMoeRoutingStrategy, URL]
    >
  ).map(([strategy, url]) => [strategy, fileURLToPath(url)])
) as Record<MiniMoeRoutingStrategy, string>;

const expertSpecs: readonly ExpertSpec[] = [
  {
    id: 'x_pos',
    signedProjection: (sample) => sample.leftInput,
    axisProjection: (sample) => sample.leftInput,
  },
  {
    id: 'x_neg',
    signedProjection: (sample) => -sample.leftInput,
    axisProjection: (sample) => sample.leftInput,
  },
  {
    id: 'y_pos',
    signedProjection: (sample) => sample.rightInput,
    axisProjection: (sample) => sample.rightInput,
  },
  {
    id: 'y_neg',
    signedProjection: (sample) => -sample.rightInput,
    axisProjection: (sample) => sample.rightInput,
  },
] as const;

function sigmoid(value: number): number {
  return 1 / (1 + Math.exp(-value));
}

function countConnectedComponents(
  nodeIds: readonly string[],
  edges: readonly {
    sourceIds: readonly string[];
    targetIds: readonly string[];
  }[]
): number {
  const neighbors = new Map<string, Set<string>>();
  for (const nodeId of nodeIds) {
    neighbors.set(nodeId, new Set<string>());
  }

  for (const edge of edges) {
    for (const sourceId of edge.sourceIds) {
      const sourceNeighbors = neighbors.get(sourceId);
      if (!sourceNeighbors) {
        continue;
      }
      for (const targetId of edge.targetIds) {
        sourceNeighbors.add(targetId);
        neighbors.get(targetId)?.add(sourceId);
      }
    }
  }

  let components = 0;
  const visited = new Set<string>();

  for (const nodeId of nodeIds) {
    if (visited.has(nodeId)) {
      continue;
    }

    components++;
    const frontier = [nodeId];
    while (frontier.length > 0) {
      const current = frontier.pop();
      if (current === undefined || visited.has(current)) {
        continue;
      }

      visited.add(current);
      for (const neighbor of neighbors.get(current) ?? []) {
        if (!visited.has(neighbor)) {
          frontier.push(neighbor);
        }
      }
    }
  }

  return Math.max(1, components);
}

function structuralBeta1(
  nodeIds: readonly string[],
  edges: readonly {
    sourceIds: readonly string[];
    targetIds: readonly string[];
  }[]
): number {
  const expandedEdgeCount = edges.reduce(
    (sum, edge) => sum + edge.sourceIds.length * edge.targetIds.length,
    0
  );

  return Math.max(
    0,
    expandedEdgeCount -
      nodeIds.length +
      countConnectedComponents(nodeIds, edges)
  );
}

function quadrantRoutingTarget(
  sample: Omit<MiniMoeRoutingSample, 'target'>
): number {
  return Math.abs(sample.leftInput) + Math.abs(sample.rightInput);
}

function buildSamples(axis: readonly number[]): MiniMoeRoutingSample[] {
  return axis.flatMap((leftInput) =>
    axis.map((rightInput) => ({
      leftInput,
      rightInput,
      target: quadrantRoutingTarget({ leftInput, rightInput }),
    }))
  );
}

function initializeExpertParameters(seed: number): ExpertParameters[] {
  const random = createDeterministicRandom(seed);
  return expertSpecs.map(() => ({
    gateWeight: (random() - 0.5) * 1.6,
    gateBias: (random() - 0.5) * 0.6,
    valueWeight: (random() - 0.5) * 1.0,
    valueBias: (random() - 0.5) * 0.4,
  }));
}

function predict(
  strategy: MiniMoeRoutingStrategy,
  parameters: readonly ExpertParameters[],
  sample: MiniMoeRoutingSample
): PredictionResult {
  const expertActivations = expertSpecs.map((expert, index) => {
    const parameter = parameters[index];
    if (!parameter) {
      throw new Error(`Missing expert parameters for index ${index}`);
    }

    const signedInput = expert.signedProjection(sample);
    const axisInput = expert.axisProjection(sample);
    const gate = sigmoid(
      parameter.gateWeight * signedInput + parameter.gateBias
    );
    const value = parameter.valueWeight * axisInput + parameter.valueBias;

    return {
      gate,
      value,
      output: gate * value,
      signedInput,
      axisInput,
    };
  });

  let selectedExpertIndex = 0;
  for (let index = 1; index < expertActivations.length; index++) {
    const selected = expertActivations[selectedExpertIndex];
    const candidate = expertActivations[index];
    if (
      candidate &&
      selected &&
      Math.abs(candidate.output) > Math.abs(selected.output)
    ) {
      selectedExpertIndex = index;
    }
  }

  if (strategy === 'linear') {
    return {
      prediction: expertActivations.reduce(
        (sum, activation) => sum + activation.output,
        0
      ),
      selectedExpertIndex,
      expertActivations,
    };
  }

  if (strategy === 'early-stop') {
    return {
      prediction: expertActivations[0]?.output ?? 0,
      selectedExpertIndex: 0,
      expertActivations,
    };
  }

  return {
    prediction: expertActivations[selectedExpertIndex]?.output ?? 0,
    selectedExpertIndex,
    expertActivations,
  };
}

function updateParameters(
  strategy: MiniMoeRoutingStrategy,
  parameters: ExpertParameters[],
  sample: MiniMoeRoutingSample,
  learningRate: number
): void {
  const result = predict(strategy, parameters, sample);
  const error = result.prediction - sample.target;

  for (let index = 0; index < parameters.length; index++) {
    if (
      strategy === 'winner-take-all' &&
      index !== result.selectedExpertIndex
    ) {
      continue;
    }
    if (strategy === 'early-stop' && index !== 0) {
      continue;
    }

    const parameter = parameters[index];
    const activation = result.expertActivations[index];
    if (!parameter || !activation) {
      continue;
    }

    const gatePrime = activation.gate * (1 - activation.gate);
    parameter.valueWeight -=
      learningRate * error * activation.gate * activation.axisInput;
    parameter.valueBias -= learningRate * error * activation.gate;
    parameter.gateWeight -=
      learningRate *
      error *
      gatePrime *
      activation.signedInput *
      activation.value;
    parameter.gateBias -= learningRate * error * gatePrime * activation.value;
  }
}

function evaluateSamples(
  strategy: MiniMoeRoutingStrategy,
  parameters: readonly ExpertParameters[],
  samples: readonly MiniMoeRoutingSample[],
  exactPredictionTolerance: number,
  dualActiveThreshold: number
): {
  readonly meanSquaredError: number;
  readonly exactWithinToleranceFraction: number;
  readonly dualActiveRegionMeanAbsoluteError: number;
} {
  if (samples.length === 0) {
    return {
      meanSquaredError: 0,
      exactWithinToleranceFraction: 0,
      dualActiveRegionMeanAbsoluteError: 0,
    };
  }

  let squaredErrorSum = 0;
  let exactCount = 0;
  let dualActiveAbsoluteErrorSum = 0;
  let dualActiveCount = 0;

  for (const sample of samples) {
    const prediction = predict(strategy, parameters, sample).prediction;
    const error = prediction - sample.target;
    squaredErrorSum += error * error;

    if (Math.abs(error) <= exactPredictionTolerance) {
      exactCount++;
    }

    if (
      Math.abs(sample.leftInput) >= dualActiveThreshold &&
      Math.abs(sample.rightInput) >= dualActiveThreshold
    ) {
      dualActiveAbsoluteErrorSum += Math.abs(error);
      dualActiveCount++;
    }
  }

  return {
    meanSquaredError: squaredErrorSum / samples.length,
    exactWithinToleranceFraction: exactCount / samples.length,
    dualActiveRegionMeanAbsoluteError:
      dualActiveCount === 0 ? 0 : dualActiveAbsoluteErrorSum / dualActiveCount,
  };
}

function summarizeSeedMetrics(
  metrics: readonly MiniMoeRoutingSeedMetrics[]
): Omit<
  MiniMoeRoutingStrategyReport,
  | 'topologyPath'
  | 'topologyNodeCount'
  | 'topologyEdgeCount'
  | 'structuralBeta1'
  | 'expertCount'
  | 'parameterCount'
  | 'samplePredictions'
> {
  const evalMeanSquaredErrors = metrics.map(
    (entry) => entry.evalMeanSquaredError
  );
  const exactFractions = metrics.map(
    (entry) => entry.exactWithinToleranceFraction
  );
  const dualActiveErrors = metrics.map(
    (entry) => entry.dualActiveRegionMeanAbsoluteError
  );

  return {
    seedMetrics: metrics,
    meanTrainMeanSquaredError: mean(
      metrics.map((entry) => entry.trainMeanSquaredError)
    ),
    stdevTrainMeanSquaredError: stdev(
      metrics.map((entry) => entry.trainMeanSquaredError)
    ),
    meanEvalMeanSquaredError: mean(evalMeanSquaredErrors),
    evalMeanSquaredErrorCi95: bootstrapMeanConfidenceInterval(
      evalMeanSquaredErrors,
      {
        seed: 0x105f00d,
      }
    ),
    stdevEvalMeanSquaredError: stdev(evalMeanSquaredErrors),
    meanExactWithinToleranceFraction: mean(exactFractions),
    exactWithinToleranceFractionCi95: bootstrapMeanConfidenceInterval(
      exactFractions,
      {
        seed: 0x201f00d,
      }
    ),
    stdevExactWithinToleranceFraction: stdev(exactFractions),
    meanDualActiveRegionMeanAbsoluteError: mean(dualActiveErrors),
    dualActiveRegionMeanAbsoluteErrorCi95: bootstrapMeanConfidenceInterval(
      dualActiveErrors,
      {
        seed: 0x301f00d,
      }
    ),
    stdevDualActiveRegionMeanAbsoluteError: stdev(dualActiveErrors),
  };
}

function samplePredictions(
  strategy: MiniMoeRoutingStrategy,
  trainedParameterSets: readonly (readonly ExpertParameters[])[],
  samplePoints: readonly (readonly [number, number])[]
): MiniMoeRoutingSamplePrediction[] {
  return samplePoints.map(([leftInput, rightInput]) => {
    const sample: MiniMoeRoutingSample = {
      leftInput,
      rightInput,
      target: quadrantRoutingTarget({ leftInput, rightInput }),
    };
    const predictions = trainedParameterSets.map(
      (parameterSet) => predict(strategy, parameterSet, sample).prediction
    );

    return {
      leftInput,
      rightInput,
      target: sample.target,
      meanPrediction: mean(predictions),
      predictionStddev: stdev(predictions),
    };
  });
}

/**
 * Handles the gnosis make Default Mini Moe Routing Config workflow.
 */
export function makeDefaultMiniMoeRoutingConfig(): MiniMoeRoutingConfig {
  return {
    label: 'gnosis-moe-routing-benchmark-v1',
    seedCount: 12,
    epochs: 1400,
    learningRate: 0.025,
    exactPredictionTolerance: 0.08,
    dualActiveThreshold: 0.4,
    trainAxis: [
      -1.5, -1.25, -1.0, -0.75, -0.5, -0.25, 0, 0.25, 0.5, 0.75, 1.0, 1.25, 1.5,
    ],
    evalAxis: [-1.4, -1.1, -0.8, -0.4, -0.1, 0.1, 0.4, 0.8, 1.1, 1.4],
    samplePoints: [
      [-1.2, -1.2],
      [-1.2, 1.2],
      [1.2, -1.2],
      [1.2, 1.2],
      [1.2, 0.0],
    ],
  };
}

async function loadTopology(
  strategy: MiniMoeRoutingStrategy
): Promise<ParsedTopology> {
  const path = DEFAULT_MINI_MOE_ROUTING_TOPOLOGY_FILES[strategy];
  const source = await readFile(path, 'utf8');
  const program = parseGgProgram(lowerUfcsSource(source));
  const foldEdges = program.edges.filter(
    (edge) => edge.type.toUpperCase() === 'FOLD'
  );
  if (foldEdges.length !== 1) {
    throw new Error(`Expected exactly one FOLD edge in ${path}`);
  }

  const foldStrategy = foldEdges[0]?.properties.strategy;
  if (foldStrategy !== strategy) {
    throw new Error(
      `Expected FOLD strategy "${strategy}" in ${path}, found "${
        foldStrategy ?? 'missing'
      }"`
    );
  }

  const expertNodes = program.nodes.filter((node) =>
    node.labels.includes('MoEExpert')
  );
  if (expertNodes.length !== 4) {
    throw new Error(`Expected exactly four MoEExpert nodes in ${path}`);
  }

  const parameterCount = expertNodes.reduce((sum, node) => {
    const rawParameters = Number(node.properties.parameters ?? 0);
    return sum + (Number.isFinite(rawParameters) ? rawParameters : 0);
  }, 0);

  return {
    strategy,
    path,
    metrics: {
      nodeCount: program.nodes.length,
      edgeCount: program.edges.length,
      structuralBeta1: structuralBeta1(
        program.nodes.map((node) => node.id),
        program.edges.map((edge) => ({
          sourceIds: edge.sourceIds,
          targetIds: edge.targetIds,
        }))
      ),
      expertCount: expertNodes.length,
      parameterCount,
    },
  };
}

async function loadDefaultTopologySuite(): Promise<
  Record<MiniMoeRoutingStrategy, ParsedTopology>
> {
  const entries = await Promise.all(
    (
      Object.keys(
        DEFAULT_MINI_MOE_ROUTING_TOPOLOGY_FILES
      ) as MiniMoeRoutingStrategy[]
    ).map(async (strategy) => [strategy, await loadTopology(strategy)] as const)
  );

  return Object.fromEntries(entries) as Record<
    MiniMoeRoutingStrategy,
    ParsedTopology
  >;
}

function assertSharedTopology(
  topologies: Record<MiniMoeRoutingStrategy, ParsedTopology>
): MiniMoeRoutingTopologyMetrics {
  const baseline = topologies.linear.metrics;
  for (const strategy of Object.keys(topologies) as MiniMoeRoutingStrategy[]) {
    const metrics = topologies[strategy].metrics;
    const isShared =
      metrics.nodeCount === baseline.nodeCount &&
      metrics.edgeCount === baseline.edgeCount &&
      metrics.structuralBeta1 === baseline.structuralBeta1 &&
      metrics.expertCount === baseline.expertCount &&
      metrics.parameterCount === baseline.parameterCount;
    if (!isShared) {
      throw new Error(
        `Topology mismatch for ${strategy}; benchmark is not parameter-matched`
      );
    }
  }

  return baseline;
}

function trainStrategy(
  strategy: MiniMoeRoutingStrategy,
  config: MiniMoeRoutingConfig,
  trainSamples: readonly MiniMoeRoutingSample[],
  evalSamples: readonly MiniMoeRoutingSample[]
): {
  readonly seedMetrics: readonly MiniMoeRoutingSeedMetrics[];
  readonly trainedParameters: readonly (readonly ExpertParameters[])[];
} {
  const seedMetrics: MiniMoeRoutingSeedMetrics[] = [];
  const trainedParameters: ExpertParameters[][] = [];

  for (let seedIndex = 0; seedIndex < config.seedCount; seedIndex++) {
    const seed = seedIndex + 1;
    const parameters = initializeExpertParameters(seed);
    const random = createDeterministicRandom((seed * 0x9e3779b1) >>> 0);

    for (let epoch = 0; epoch < config.epochs; epoch++) {
      const orderedSamples = [...trainSamples];
      shuffleInPlace(orderedSamples, random);
      for (const sample of orderedSamples) {
        updateParameters(strategy, parameters, sample, config.learningRate);
      }
    }

    const trainEvaluation = evaluateSamples(
      strategy,
      parameters,
      trainSamples,
      config.exactPredictionTolerance,
      config.dualActiveThreshold
    );
    const evalEvaluation = evaluateSamples(
      strategy,
      parameters,
      evalSamples,
      config.exactPredictionTolerance,
      config.dualActiveThreshold
    );

    seedMetrics.push({
      seed,
      trainMeanSquaredError: trainEvaluation.meanSquaredError,
      evalMeanSquaredError: evalEvaluation.meanSquaredError,
      exactWithinToleranceFraction: evalEvaluation.exactWithinToleranceFraction,
      dualActiveRegionMeanAbsoluteError:
        evalEvaluation.dualActiveRegionMeanAbsoluteError,
    });
    trainedParameters.push(parameters.map((parameter) => ({ ...parameter })));
  }

  return {
    seedMetrics,
    trainedParameters,
  };
}

/**
 * Handles the gnosis run Gnosis Mini Moe Routing Benchmark workflow.
 */
export async function runGnosisMiniMoeRoutingBenchmark(
  config: MiniMoeRoutingConfig = makeDefaultMiniMoeRoutingConfig()
): Promise<GnosisMiniMoeRoutingBenchmarkReport> {
  const topologies = await loadDefaultTopologySuite();
  const sharedTopology = assertSharedTopology(topologies);
  const trainSamples = buildSamples(config.trainAxis);
  const evalSamples = buildSamples(config.evalAxis);

  const strategies = {} as Record<
    MiniMoeRoutingStrategy,
    MiniMoeRoutingStrategyReport
  >;
  for (const strategy of Object.keys(topologies) as MiniMoeRoutingStrategy[]) {
    const trained = trainStrategy(strategy, config, trainSamples, evalSamples);
    const summary = summarizeSeedMetrics(trained.seedMetrics);
    strategies[strategy] = {
      topologyPath: topologies[strategy].path,
      topologyNodeCount: topologies[strategy].metrics.nodeCount,
      topologyEdgeCount: topologies[strategy].metrics.edgeCount,
      structuralBeta1: topologies[strategy].metrics.structuralBeta1,
      expertCount: topologies[strategy].metrics.expertCount,
      parameterCount: topologies[strategy].metrics.parameterCount,
      ...summary,
      samplePredictions: samplePredictions(
        strategy,
        trained.trainedParameters,
        config.samplePoints
      ),
    };
  }

  const rankingByEvalMeanSquaredError = (
    Object.keys(strategies) as MiniMoeRoutingStrategy[]
  ).sort(
    (left, right) =>
      strategies[left].meanEvalMeanSquaredError -
      strategies[right].meanEvalMeanSquaredError
  );

  return {
    label: config.label,
    config: {
      seedCount: config.seedCount,
      epochs: config.epochs,
      learningRate: config.learningRate,
      exactPredictionTolerance: config.exactPredictionTolerance,
      dualActiveThreshold: config.dualActiveThreshold,
      trainSampleCount: trainSamples.length,
      evalSampleCount: evalSamples.length,
      targetFamily: 'quadrant-l1-routing',
    },
    topology: sharedTopology,
    strategies,
    rankingByEvalMeanSquaredError,
    predictedRankingMatches:
      rankingByEvalMeanSquaredError[0] === 'linear' &&
      rankingByEvalMeanSquaredError[1] === 'winner-take-all' &&
      rankingByEvalMeanSquaredError[2] === 'early-stop',
  };
}

function formatSamplePair(sample: MiniMoeRoutingSamplePrediction): string {
  return `(${sample.leftInput.toFixed(2)}, ${sample.rightInput.toFixed(2)})`;
}

function formatInterval(interval: ConfidenceInterval): string {
  return `[${interval.low.toFixed(3)}, ${interval.high.toFixed(3)}]`;
}

/**
 * Handles the gnosis render Gnosis Mini Moe Routing Benchmark Markdown workflow.
 */
export function renderGnosisMiniMoeRoutingBenchmarkMarkdown(
  report: GnosisMiniMoeRoutingBenchmarkReport
): string {
  const lines: string[] = [];
  lines.push('# Gnosis Mini-MoE Routing Benchmark');
  lines.push('');
  lines.push(`- Label: \`${report.label}\``);
  lines.push(`- Target family: \`${report.config.targetFamily}\``);
  lines.push(`- Seeds: \`${report.config.seedCount}\``);
  lines.push(`- Epochs: \`${report.config.epochs}\``);
  lines.push(`- Learning rate: \`${report.config.learningRate}\``);
  lines.push(`- Train samples: \`${report.config.trainSampleCount}\``);
  lines.push(`- Eval samples: \`${report.config.evalSampleCount}\``);
  lines.push(`- Shared parameter count: \`${report.topology.parameterCount}\``);
  lines.push(`- Shared structural β₁: \`${report.topology.structuralBeta1}\``);
  lines.push(
    `- Dual-active threshold: \`${report.config.dualActiveThreshold}\``
  );
  lines.push(
    `- Predicted ranking recovered: \`${
      report.predictedRankingMatches ? 'yes' : 'no'
    }\``
  );
  lines.push('');
  lines.push('## Aggregated Metrics');
  lines.push('');
  lines.push(
    '| Strategy | Mean eval MSE | Eval MSE 95% CI | Exact-within-tolerance | Exact 95% CI | Dual-active abs error | Dual-active 95% CI |'
  );
  lines.push('|---|---:|---:|---:|---:|---:|---:|');

  for (const strategy of report.rankingByEvalMeanSquaredError) {
    const metrics = report.strategies[strategy];
    lines.push(
      `| \`${strategy}\` | ${metrics.meanEvalMeanSquaredError.toFixed(
        3
      )} | ${formatInterval(
        metrics.evalMeanSquaredErrorCi95
      )} | ${metrics.meanExactWithinToleranceFraction.toFixed(
        3
      )} | ${formatInterval(
        metrics.exactWithinToleranceFractionCi95
      )} | ${metrics.meanDualActiveRegionMeanAbsoluteError.toFixed(
        3
      )} | ${formatInterval(metrics.dualActiveRegionMeanAbsoluteErrorCi95)} |`
    );
  }

  lines.push('');
  lines.push('## Sample Predictions');
  lines.push('');
  lines.push(
    '| Strategy | Input pair | Target | Mean prediction | Prediction σ |'
  );
  lines.push('|---|---|---:|---:|---:|');

  for (const strategy of report.rankingByEvalMeanSquaredError) {
    for (const sample of report.strategies[strategy].samplePredictions) {
      lines.push(
        `| \`${strategy}\` | ${formatSamplePair(
          sample
        )} | ${sample.target.toFixed(3)} | ${sample.meanPrediction.toFixed(
          3
        )} | ${sample.predictionStddev.toFixed(3)} |`
      );
    }
  }

  lines.push('');
  lines.push(
    'Interpretation: four sign-specialized experts share one routed topology and a fixed 16-parameter budget. The linear fold can aggregate the two active experts required for `|x| + |y|`, while winner-take-all and early-stop leave a persistent dual-path under-recombination floor.'
  );
  lines.push('');
  return `${lines.join('\n')}\n`;
}
