// @a0n/gnosis/src/benchmarks/topological-xor-skip-benchmark.ts
import {
  decideTopologicalXorSkip,
  issueTopologicalXorCertificate,
  planTopologicalXorSkips,
  type TopologicalXorBoundary,
} from '../topological-xor-certificate';

export interface TopologicalXorSkipBenchmarkConfig {
  readonly candidates: number;
  readonly iterations: number;
  readonly retentionBudget: number;
}

export interface TopologicalXorSkipBenchmarkReport {
  readonly label: 'gnosis-topological-xor-skip-benchmark-v1';
  readonly candidates: number;
  readonly iterations: number;
  readonly retentionBudget: number;
  readonly elapsedMs: number;
  readonly meanIterationMs: number;
  readonly decisionsPerSecond: number;
  readonly skipTargets: number;
  readonly retainedSkipTargets: number;
  readonly recomputableSkipTargets: number;
  readonly measureTargets: number;
  readonly reconstructTargets: number;
  readonly skipRatePerthou: number;
}

function candidateBoundaries(index: number): readonly TopologicalXorBoundary[] {
  const label = `boundary-${index}`;
  if (index % 4 === 0) {
    return [{ label, status: 'measured' }];
  }
  if (index % 4 === 1) {
    return [
      { label, status: 'projected' },
      { label, status: 'projected' },
    ];
  }
  return [
    { label, status: 'measured' },
    { label, status: 'measured' },
  ];
}

/**
 * Handles the gnosis run Topological Xor Skip Benchmark workflow.
 */
export function runTopologicalXorSkipBenchmark(
  config: Partial<TopologicalXorSkipBenchmarkConfig> = {}
): TopologicalXorSkipBenchmarkReport {
  const candidates = config.candidates ?? 10_000;
  const iterations = config.iterations ?? 100;
  const retentionBudget = config.retentionBudget ?? Math.floor(candidates / 4);
  if (!Number.isSafeInteger(candidates) || candidates <= 0) {
    throw new Error('candidates must be a positive safe integer');
  }
  if (!Number.isSafeInteger(iterations) || iterations <= 0) {
    throw new Error('iterations must be a positive safe integer');
  }
  if (!Number.isSafeInteger(retentionBudget) || retentionBudget < 0) {
    throw new Error('retentionBudget must be a non-negative safe integer');
  }

  let lastPlan = planTopologicalXorSkips({ decisions: [] });
  const started = performance.now();
  for (let iteration = 0; iteration < iterations; iteration += 1) {
    const decisions = Array.from({ length: candidates }, (_, index) => {
      const certificate = issueTopologicalXorCertificate({
        certificateId: `txor-bench-${iteration}-${index}`,
        target: `runtime://skip-bench/${index}`,
        boundaries: candidateBoundaries(index),
      });
      return decideTopologicalXorSkip({
        certificate,
        allowProjectedSkip: index % 8 === 1,
      });
    });
    const priorityByTarget: Record<string, number> = {};
    for (let index = 0; index < candidates; index += 1) {
      priorityByTarget[`runtime://skip-bench/${index}`] =
        (candidates - index) * ((index % 7) + 1);
    }
    lastPlan = planTopologicalXorSkips({
      decisions,
      retentionBudget,
      priorityByTarget,
    });
  }
  const elapsedMs = performance.now() - started;
  const totalDecisions = candidates * iterations;
  const skipTargets = lastPlan.skipTargets.length;

  return {
    label: 'gnosis-topological-xor-skip-benchmark-v1',
    candidates,
    iterations,
    retentionBudget,
    elapsedMs,
    meanIterationMs: elapsedMs / iterations,
    decisionsPerSecond: (totalDecisions / elapsedMs) * 1000,
    skipTargets,
    retainedSkipTargets: lastPlan.retainedSkipTargets.length,
    recomputableSkipTargets: lastPlan.recomputableSkipTargets.length,
    measureTargets: lastPlan.measureTargets.length,
    reconstructTargets: lastPlan.reconstructTargets.length,
    skipRatePerthou: Math.floor((skipTargets * 1000) / candidates),
  };
}

function argNumber(name: string): number | undefined {
  const prefix = `--${name}=`;
  const arg = process.argv.find((item) => item.startsWith(prefix));
  return arg === undefined ? undefined : Number(arg.slice(prefix.length));
}

if (process.argv[1]?.endsWith('topological-xor-skip-benchmark.ts')) {
  console.log(
    JSON.stringify(
      runTopologicalXorSkipBenchmark({
        candidates: argNumber('candidates'),
        iterations: argNumber('iterations'),
        retentionBudget: argNumber('retention-budget'),
      }),
      null,
      2
    )
  );
}
