// @a0n/gnosis/src/benchmarks/x-gnosis-vm-benchmark.ts
import { FlowFrame } from '../packed-knot-to-flowframe';
import { XGnosisEdgeVM } from '../x-gnosis-vm';


// If statistics.js is unavailable, we just emit basic timing logic gracefully. 
// A primitive timer in case statistics is purely UI oriented.
class BenchmarkClock {
  private start = performance.now();
  stop(label: string) {
    const elapsed = performance.now() - this.start;
    console.log(`[X-Gnosis Benchmark] ${label} || ${elapsed.toFixed(3)}ms`);
  }
}

/**
 * X-Gnosis VM Semantic Validation Benchmark
 * Verifies execution of 1,000,000 frames assessing thermodynamic exhaustion natively.
 */
async function runThroughputBenchmark() {
  console.log("=== X-Gnosis VM Thermodynamic Bound Benchmark ===");

  const N = 1_000_000;
  // Synthesize 1M flowframes (lightweight)
  const syntheticPayload = new Uint8Array(1); 
  const frames: FlowFrame[] = new Array(N);
  for (let i = 0; i < N; i++) {
    frames[i] = {
      streamId: i % 100,
      sequence: i,
      flags: 0,
      payload: syntheticPayload
    };
  }

  // 1. Unconstrained Baseline
  const clock1 = new BenchmarkClock();
  let v_raw = 0;
  for (let i = 0; i < N; i++) {
    // Basic loop just iterating and incrementing
    const f = frames[i];
    v_raw += 1;
  }
  clock1.stop(`Raw Baseline V8 Loop (${N} frames)`);

  // 2. Thermodynamic VM Firewall Constraint
  const vm = new XGnosisEdgeVM(1_000_000, 0); // Large budget to accommodate N observations
  const clock2 = new BenchmarkClock();
  for (let i = 0; i < N; i++) {
    vm.processFrame(frames[i]);
  }
  clock2.stop(`X-Gnosis Edge VM Checked Execution (${N} frames)`);

  // 3. Event Horizon Halt Testing Boundary
  // Re-run with a tiny 10k budget to measure branch prediction impact when halted
  const haltingVM = new XGnosisEdgeVM(10_000, 0);
  let haltedAt = 0;
  const clock3 = new BenchmarkClock();
  for (let i = 0; i < N; i++) {
    try {
      haltingVM.processFrame(frames[i]);
    } catch (e: any) {
      if (e.message.includes('Thermodynamic Halting')) {
        haltedAt = i;
        break; // Prove halting bounds escape the loop seamlessly
      }
    }
  }
  clock3.stop(`Thermodynamic Halting Trigger (${haltedAt} frames evaluated)`);

  console.log("==================================================");
}

// Execute independently when triggered directly by bun/node
if (typeof process !== 'undefined' && process.argv[1] && process.argv[1].endsWith('x-gnosis-vm-benchmark.ts')) {
    runThroughputBenchmark().catch(console.error);
}

export { runThroughputBenchmark };
