// @a0n/gnosis/src/crdt/index.ts
'use aeon';

export { QDoc, QMap, QArray, QText, QCounter } from './qdoc';
export { QCorridor } from './qcorridor';
export type {
  QDocOptions,
  QDocDelta,
  QDocDeltaNode,
  QDocDeltaEdge,
  QDocUpdateHandler,
  QDocObserveHandler,
  QDocEvent,
  QMapKeyChange,
  QMapEvent,
} from './qdoc';
export type {
  QCorridorOptions,
  QCorridorExecutionOptions,
  QCorridorMode,
  QCorridorReuseScope,
  QCorridorSession,
  QCorridorMetrics,
  QCorridorEntry,
  QCorridorRecord,
  QCorridorEvent,
  QCorridorEvidence,
  QCorridorObservation,
  QCorridorObservationRole,
  QCorridorObservationStatus,
  QCorridorLookup,
} from './qcorridor';
export {
  QDocAeonRelay,
  QDocRelay,
  createQDocAeonRelayJoinEnvelope,
} from './aeon-relay';
export type {
  QDocAeonRelayAttributes,
  QDocAeonRelayAttributeValue,
  QDocAeonRelayConfig,
  QDocAeonRelayJoinEnvelope,
  QDocAeonRelayReadyStrategy,
  QDocAeonRelaySpan,
  QDocAeonRelayStatus,
  QDocAeonRelayTelemetry,
  QDocAeonRelayTelemetryEvent,
  QDocAeonRelayTelemetryStage,
  QDocRelayAttributes,
  QDocRelayAttributeValue,
  QDocRelayConfig,
  QDocRelayReadyStrategy,
  QDocRelaySpan,
  QDocRelayStatus,
  QDocRelayTelemetry,
  QDocRelayTelemetryEvent,
  QDocRelayTelemetryStage,
} from './aeon-relay';
export {
  Doc,
  Map,
  Array,
  Text,
  XmlFragment,
  XmlElement,
  UndoManager,
  applyUpdate,
  encodeStateAsUpdate,
  transact,
  encodeStateVector,
  diffUpdate,
  mergeUpdates,
} from './yjs-compat';
