// @a0n/gnosis/forkjoin-app/client/components/Sections/DashRelaySection.tsx
/**
 * DashRelaySection - Crypto Relay Layer
 *
 * Managed crypto relay infrastructure combining Dash storage + Relay transport
 * with UCAN identity, zero-knowledge encryption, and CRDT brokering:
 * - UCAN Identity (did:key, capability delegation)
 * - Zero-Knowledge Encryption (ECIES-P256, AES-256-GCM)
 * - Privacy-First Relay (no plaintext on server)
 * - Global Edge Network (300+ PoPs, <50ms)
 * - CRDT Sync (Yjs, Automerge compatible)
 * - Simple Integration (npm install, 3 lines)
 */

import {
  Zap,
  Fingerprint,
  Lock,
  Shield,
  Globe,
  Code,
  Network,
  RefreshCw,
  Key,
  FileKey,
  Users,
  Layers,
  Timer,
  CloudOff,
  Server,
  Cpu,
} from '@a0n/aeon-icons';
import {
  DASHRELAY_TRUST_LOG_DESCRIPTION,
  TERM_CRYPTO_RELAY,
  TERM_DASHRELAY,
} from '../../../src/terminology';

const features = [
  // Row 1: Identity & Auth
  {
    icon: Fingerprint,
    title: 'UCAN Identity',
    metric: 'did:key',
    description:
      'Decentralized identifiers with ECDSA P-256. No central authority. User-owned keys.',
    category: 'identity',
  },
  {
    icon: Key,
    title: 'Capability Chains',
    metric: 'Delegated',
    description:
      'UCAN capability delegation. Grant scoped access without a central server.',
    category: 'identity',
  },
  {
    icon: Shield,
    title: 'Zero-Trust Auth',
    metric: 'Verify all',
    description:
      'Every message cryptographically signed. No implicit trust. Verify everything.',
    category: 'identity',
  },
  {
    icon: FileKey,
    title: 'Key Rotation',
    metric: 'Seamless',
    description:
      'Rotate keys without losing access. Forward secrecy via ephemeral ECDH.',
    category: 'identity',
  },

  // Row 2: Encryption
  {
    icon: Lock,
    title: 'ZK Encryption',
    metric: 'E2E',
    description:
      'ECIES-P256 + AES-256-GCM. End-to-end encrypted. Server never sees plaintext.',
    category: 'encryption',
  },
  {
    icon: CloudOff,
    title: 'No Plaintext Relay',
    metric: 'Private',
    description:
      'Server relays opaque ciphertext only. Zero knowledge of document content.',
    category: 'encryption',
  },
  {
    icon: RefreshCw,
    title: 'Forward Secrecy',
    metric: 'Ephemeral',
    description:
      'Ephemeral ECDH key exchange. Compromise one session, others stay safe.',
    category: 'encryption',
  },
  {
    icon: Layers,
    title: 'Encrypted CRDTs',
    metric: 'Merge',
    description:
      'CRDTs encrypted at rest and in transit. Conflict-free sync without exposure.',
    category: 'encryption',
  },

  // Row 3: Sync & Transport
  {
    icon: Network,
    title: 'WebTransport',
    metric: 'UDP',
    description:
      'UDP-based relay with automatic WebSocket fallback. Sub-10ms latency.',
    category: 'sync',
  },
  {
    icon: Users,
    title: 'Yjs + Automerge',
    metric: 'CRDT',
    description:
      'Native providers for both Yjs and Automerge. connect() or connectAutomerge(). Real-time collaborative editing.',
    category: 'sync',
  },
  {
    icon: Timer,
    title: 'Presence',
    metric: 'Real-time',
    description:
      'Built-in cursor tracking, selection awareness, and user presence channels.',
    category: 'sync',
  },
  {
    icon: Zap,
    title: 'Offline-First',
    metric: 'Queue',
    description:
      'Changes queued locally when offline. Auto-sync on reconnect with conflict resolution.',
    category: 'sync',
  },

  // Row 4: Infrastructure
  {
    icon: Globe,
    title: 'Global Edge',
    metric: '300+ PoPs',
    description:
      'Cloudflare edge network. Sub-50ms global fanout. Data stays close to users.',
    category: 'infra',
  },
  {
    icon: Server,
    title: 'DO Hibernation',
    metric: '$0 idle',
    description:
      'Durable Object rooms hibernate when empty. Pay only for active connections.',
    category: 'infra',
  },
  {
    icon: Cpu,
    title: 'Edge Compute',
    metric: 'V8',
    description:
      'Auth validation and routing at the edge. No origin round-trip. Instant.',
    category: 'infra',
  },
  {
    icon: Code,
    title: '3-Line Setup',
    metric: 'Simple',
    description:
      'npm install, import, connect. Production-ready sync in minutes, not months.',
    category: 'infra',
  },
];

const categoryColors: Record<string, string> = {
  identity: '#10b981',
  encryption: '#8b5cf6',
  sync: '#06b6d4',
  infra: '#3b82f6',
};

/**
 * Renders the Dash Relay Section view.
 */
export function DashRelaySection() {
  return (
    <section
      id="dashrelay"
      className="section dashrelay-section"
      aria-labelledby="dashrelay-title"
    >
      <div className="container">
        <div className="section-header">
          <span className="layer-label">{TERM_CRYPTO_RELAY}</span>
          <h2 id="dashrelay-title" className="product-section-title">
            {TERM_DASHRELAY}
          </h2>
          <p className="section-description">
            {DASHRELAY_TRUST_LOG_DESCRIPTION}
          </p>
        </div>

        <div className="dashrelay-grid">
          {features.map((feature) => (
            <div
              key={feature.title}
              className="dashrelay-tile"
              style={
                {
                  '--tile-color': categoryColors[feature.category],
                } as React.CSSProperties
              }
            >
              <feature.icon size={24} className="tile-icon" />
              <div className="tile-content">
                <div className="tile-header">
                  <h3>{feature.title}</h3>
                  <span className="tile-metric">{feature.metric}</span>
                </div>
                <p>{feature.description}</p>
              </div>
            </div>
          ))}
        </div>

        <div className="comparison-table">
          <h3>vs Other Sync Platforms</h3>
          <table className="comparison-grid-table">
            <thead>
              <tr>
                <th>Feature</th>
                <th>dashrelay</th>
                <th>Liveblocks</th>
                <th>Partykit</th>
                <th>Socket.io</th>
                <th>Firebase</th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>UCAN Identity</td>
                <td className="yes">&#10003; Native</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
              </tr>
              <tr>
                <td>E2E Encryption</td>
                <td className="yes">&#10003; ZK</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
              </tr>
              <tr>
                <td>CRDT Sync</td>
                <td className="yes">&#10003; Yjs + Automerge</td>
                <td className="yes">&#10003; Yes</td>
                <td className="partial">Basic</td>
                <td className="no">No</td>
                <td className="no">No</td>
              </tr>
              <tr>
                <td>WebTransport</td>
                <td className="yes">&#10003; Native</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
              </tr>
              <tr>
                <td>Zero Idle Cost</td>
                <td className="yes">&#10003; DO Hibernate</td>
                <td className="no">No</td>
                <td className="yes">&#10003; Yes</td>
                <td className="no">No</td>
                <td className="no">No</td>
              </tr>
              <tr>
                <td>Global Edge</td>
                <td className="yes">&#10003; 300+ PoPs</td>
                <td className="partial">Limited</td>
                <td className="yes">&#10003; CF</td>
                <td className="no">No</td>
                <td className="partial">Regions</td>
              </tr>
              <tr>
                <td>Self-Hosted</td>
                <td className="yes">&#10003; Yes</td>
                <td className="no">No</td>
                <td className="yes">&#10003; Yes</td>
                <td className="yes">&#10003; Yes</td>
                <td className="no">No</td>
              </tr>
              <tr>
                <td>No Vendor Lock-in</td>
                <td className="yes">&#10003; Open</td>
                <td className="no">Proprietary</td>
                <td className="partial">Partial</td>
                <td className="yes">&#10003; Open</td>
                <td className="no">Proprietary</td>
              </tr>
            </tbody>
          </table>
        </div>

        <div className="cta-row">
          <a href="#/dashrelay" className="button button-primary">
            Explore dashrelay
          </a>
          <a
            href="https://dashrelay.com"
            className="button button-secondary"
            target="_blank"
            rel="noopener noreferrer"
          >
            Visit dashrelay.com
          </a>
        </div>
      </div>

      <style>{`
        .dashrelay-section {
          background: linear-gradient(180deg, var(--color-bg) 0%, #0a0f14 100%);
        }

        .section-header {
          text-align: center;
          margin-bottom: var(--space-3xl);
        }

        .section-description {
          font-size: 1.125rem;
          color: var(--color-text-secondary);
          max-width: 600px;
          margin: var(--space-md) auto 0;
          line-height: 1.6;
        }

        .dashrelay-grid {
          display: grid;
          grid-template-columns: repeat(4, 1fr);
          gap: var(--space-md);
          margin-bottom: var(--space-3xl);
        }

        .dashrelay-tile {
          display: flex;
          gap: var(--space-md);
          padding: var(--space-md);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 10px;
          transition: all 0.2s ease;
        }

        .dashrelay-tile:hover {
          border-color: var(--tile-color);
          transform: translateY(-2px);
        }

        .tile-icon {
          color: var(--tile-color);
          flex-shrink: 0;
          margin-top: 2px;
        }

        .tile-content {
          flex: 1;
          min-width: 0;
        }

        .tile-header {
          display: flex;
          align-items: center;
          justify-content: space-between;
          gap: var(--space-sm);
          margin-bottom: var(--space-xs);
        }

        .tile-header h3 {
          font-size: 0.9rem;
          font-weight: 600;
          color: var(--color-text);
          margin: 0;
          white-space: nowrap;
          overflow: hidden;
          text-overflow: ellipsis;
        }

        .dashrelay-tile .tile-metric {
          font-family: var(--font-mono);
          font-size: 0.7rem;
          font-weight: 600;
          color: #e2e8f0;
          padding: 2px 6px;
          background: #1a1a24;
          border-radius: 4px;
          flex-shrink: 0;
        }

        .dashrelay-tile p {
          font-size: 0.8rem;
          color: var(--color-text-muted);
          margin: 0;
          line-height: 1.4;
        }

        .comparison-table {
          max-width: 900px;
          margin: 0 auto var(--space-3xl);
          text-align: center;
        }

        .comparison-table h3 {
          margin-bottom: var(--space-lg);
          color: var(--color-text-muted);
          text-transform: uppercase;
          letter-spacing: 0.1em;
          font-size: 0.85rem;
        }

        .comparison-grid-table {
          width: 100%;
          border-collapse: collapse;
          border-radius: 8px;
          overflow: hidden;
          background: var(--color-bg-tertiary);
        }

        .comparison-grid-table th,
        .comparison-grid-table td {
          padding: var(--space-sm) var(--space-md);
          font-size: 0.8rem;
          border: 1px solid var(--color-bg-tertiary);
        }

        .comparison-grid-table th {
          background: var(--color-bg-secondary);
          font-weight: 600;
          text-align: center;
        }

        .comparison-grid-table td {
          background: var(--color-bg);
        }

        .comparison-grid-table td:first-child {
          text-align: left;
          color: var(--color-text-secondary);
        }

        .comparison-grid-table td:not(:first-child) {
          text-align: center;
          font-weight: 600;
        }

        .comparison-grid-table .yes {
          color: #10b981;
        }

        .comparison-grid-table .no {
          color: var(--color-text-muted);
        }

        .comparison-grid-table .partial {
          color: #f59e0b;
        }

        .cta-row {
          display: flex;
          justify-content: center;
          gap: var(--space-md);
        }

        @media (max-width: 1200px) {
          .dashrelay-grid {
            grid-template-columns: repeat(2, 1fr);
          }
        }

        @media (max-width: 600px) {
          .dashrelay-grid {
            grid-template-columns: 1fr;
          }

          .dashrelay-tile {
            flex-direction: column;
            text-align: center;
          }

          .tile-header {
            justify-content: center;
          }

          .tile-icon {
            margin: 0 auto;
          }

          .comparison-table {
            overflow-x: auto;
          }

          .comparison-grid-table th,
          .comparison-grid-table td {
            padding: var(--space-xs) var(--space-sm);
            font-size: 0.7rem;
          }
        }
      `}</style>
    </section>
  );
}
