// @a0n/gnosis/forkjoin-app/client/components/Sections/DashSection.tsx
/**
 * DashSection - The Local-First, AI-Native Database
 *
 * Shows the incredible capabilities from the Dash README:
 * - Production-Grade Storage (OPFS, SQLite, ACID)
 * - Hardware-Accelerated AI (Vector Search, WebNN, WebGPU)
 * - Modern Sync (WebTransport, UCANs, Lens)
 * - Graphics Ready (Zero-copy, R-Tree)
 * - Firebase-Compatible API (Drop-in replacement)
 */

import {
  Database,
  HardDrive,
  Archive,
  Layers,
  Cpu,
  Brain,
  Zap,
  Search,
  RefreshCw,
  Shield,
  GitBranch,
  Flame,
  Box,
  MapPin,
  Wifi,
  Lock,
} from '@a0n/aeon-icons';

const features = [
  // Row 1: Core Storage
  {
    icon: Database,
    title: 'SQLite over OPFS',
    metric: '1ms',
    description:
      'Full ACID transactions on multi-GB SQLite in the browser. Near-native performance.',
    category: 'storage',
  },
  {
    icon: HardDrive,
    title: 'Zero-Copy IO',
    metric: 'BYOB',
    description:
      'Bring Your Own Buffer readers for maximum throughput. No serialization overhead.',
    category: 'storage',
  },
  {
    icon: Archive,
    title: 'Stateful Serverless',
    metric: '$0',
    description:
      'Cloudflare Durable Objects with native SQLite. Zero idle cost with hibernation.',
    category: 'storage',
  },
  {
    icon: Layers,
    title: 'Smart Tiering',
    metric: 'Auto',
    description:
      'KV (1ms) → D1 (5ms) → R2 (10ms). Automatic data placement by access pattern.',
    category: 'storage',
  },

  // Row 2: AI Features
  {
    icon: Search,
    title: 'Vector Search',
    metric: 'Native',
    description:
      'Built-in vector embeddings. Semantic search by meaning, not just keywords.',
    category: 'ai',
  },
  {
    icon: Cpu,
    title: 'WebNN Acceleration',
    metric: 'NPU',
    description:
      'Apple Neural Engine, Qualcomm Hexagon. Hardware AI on any device.',
    category: 'ai',
  },
  {
    icon: Zap,
    title: 'WebGPU Fallback',
    metric: '1000x',
    description:
      'High-performance GPU parallelization when NPU unavailable. Massive speedup.',
    category: 'ai',
  },
  {
    icon: Brain,
    title: 'Semantic Queries',
    metric: 'Smart',
    description:
      'Find data by meaning at native speeds. "breakfast" matches "eggs and milk".',
    category: 'ai',
  },

  // Row 3: Sync & Transport
  {
    icon: Wifi,
    title: 'WebTransport',
    metric: 'UDP',
    description:
      'High-frequency streams for cursors, games, real-time. UDP-like performance.',
    category: 'sync',
  },
  {
    icon: RefreshCw,
    title: 'DO Hibernation',
    metric: '$0 idle',
    description:
      'WebSocket connections that sleep when idle. Pay only for active use.',
    category: 'sync',
  },
  {
    icon: Shield,
    title: 'Zero-Trust Auth',
    metric: 'UCANs',
    description:
      'Decentralized authorization. User-controlled capabilities. No central server.',
    category: 'sync',
  },
  {
    icon: GitBranch,
    title: 'Lens Migrations',
    metric: 'Bi-dir',
    description:
      'Bidirectional schema migrations. Infinite backward compatibility.',
    category: 'sync',
  },

  // Row 4: Graphics & Firebase
  {
    icon: Box,
    title: 'Direct Buffers',
    metric: 'Zero-copy',
    description:
      'Pipe data directly into Three.js. No serialization. Just raw performance.',
    category: 'graphics',
  },
  {
    icon: MapPin,
    title: 'Spatial Indexing',
    metric: 'R-Tree',
    description:
      '3D bounding box queries for massive visualizations. Instant spatial lookups.',
    category: 'graphics',
  },
  {
    icon: Flame,
    title: 'Firebase Drop-In',
    metric: '150+ APIs',
    description:
      'Same imports, same code, 10-50x faster. Firestore, Auth, Realtime, Storage.',
    category: 'firebase',
  },
  {
    icon: Lock,
    title: 'E2E Encrypted',
    metric: 'Private',
    description:
      'Your data stays yours. Encrypted at rest, in transit, everywhere.',
    category: 'firebase',
  },
];

const categoryColors: Record<string, string> = {
  storage: '#f59e0b',
  ai: '#10b981',
  sync: '#06b6d4',
  graphics: '#3b82f6',
  firebase: '#ef4444',
};

/**
 * Renders the Dash Section view.
 */
export function DashSection() {
  return (
    <section
      id="dash"
      className="section dash-section"
      aria-labelledby="dash-title"
    >
      <div className="container">
        <div className="section-header">
          <span className="layer-label">Storage Layer</span>
          <h2 id="dash-title" className="product-section-title">
            Dash
          </h2>
          <p className="section-description">
            The local-first, AI-native database for the modern web.
          </p>
        </div>

        <div className="dash-grid">
          {features.map((feature) => (
            <div
              key={feature.title}
              className="dash-tile"
              style={
                {
                  '--tile-color': categoryColors[feature.category],
                } as React.CSSProperties
              }
            >
              <feature.icon size={24} className="tile-icon" />
              <div className="tile-content">
                <div className="tile-header">
                  <h3>{feature.title}</h3>
                  <span className="tile-metric">{feature.metric}</span>
                </div>
                <p>{feature.description}</p>
              </div>
            </div>
          ))}
        </div>

        <div className="comparison-table">
          <h3>vs Other Databases</h3>
          <table className="comparison-grid-table">
            <thead>
              <tr>
                <th>Feature</th>
                <th>Dash</th>
                <th>IndexedDB</th>
                <th>PouchDB</th>
                <th>Dexie</th>
                <th>Firebase</th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>SQLite (ACID)</td>
                <td className="yes">✓ Yes</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
              </tr>
              <tr>
                <td>OPFS Storage</td>
                <td className="yes">✓ Yes</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="partial">Partial</td>
                <td className="no">No</td>
              </tr>
              <tr>
                <td>Vector Search</td>
                <td className="yes">✓ Native</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="partial">Extension</td>
              </tr>
              <tr>
                <td>WebNN/WebGPU</td>
                <td className="yes">✓ Yes</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
              </tr>
              <tr>
                <td>Zero-Copy Buffers</td>
                <td className="yes">✓ Yes</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
              </tr>
              <tr>
                <td>Offline-First</td>
                <td className="yes">✓ Yes</td>
                <td className="yes">✓ Yes</td>
                <td className="yes">✓ Yes</td>
                <td className="yes">✓ Yes</td>
                <td className="partial">Partial</td>
              </tr>
              <tr>
                <td>Real-time Sync</td>
                <td className="yes">✓ WebTransport</td>
                <td className="no">No</td>
                <td className="yes">✓ CouchDB</td>
                <td className="no">No</td>
                <td className="yes">✓ Yes</td>
              </tr>
              <tr>
                <td>Schema Migrations</td>
                <td className="yes">✓ Lens</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="partial">Basic</td>
                <td className="no">No</td>
              </tr>
              <tr>
                <td>E2E Encryption</td>
                <td className="yes">✓ Yes</td>
                <td className="no">No</td>
                <td className="partial">Plugin</td>
                <td className="no">No</td>
                <td className="no">No</td>
              </tr>
              <tr>
                <td>Zero Idle Cost</td>
                <td className="yes">✓ DO Hibernate</td>
                <td className="yes">✓ Local</td>
                <td className="yes">✓ Local</td>
                <td className="yes">✓ Local</td>
                <td className="no">No</td>
              </tr>
            </tbody>
          </table>
        </div>

        <div className="cta-row">
          <a href="#/dash" className="button button-primary">
            Explore Dash
          </a>
        </div>
      </div>

      <style>{`
        .dash-section {
          background: linear-gradient(180deg, var(--color-bg-secondary) 0%, var(--color-bg) 100%);
        }

        .section-header {
          text-align: center;
          margin-bottom: var(--space-3xl);
        }

        .section-description {
          font-size: 1.125rem;
          color: var(--color-text-secondary);
          max-width: 600px;
          margin: var(--space-md) auto 0;
          line-height: 1.6;
        }

        .dash-grid {
          display: grid;
          grid-template-columns: repeat(4, 1fr);
          gap: var(--space-md);
          margin-bottom: var(--space-3xl);
        }

        .dash-tile {
          display: flex;
          gap: var(--space-md);
          padding: var(--space-md);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 10px;
          transition: all 0.2s ease;
        }

        .dash-tile:hover {
          border-color: var(--tile-color);
          transform: translateY(-2px);
        }

        .tile-icon {
          color: var(--tile-color);
          flex-shrink: 0;
          margin-top: 2px;
        }

        .tile-content {
          flex: 1;
          min-width: 0;
        }

        .tile-header {
          display: flex;
          align-items: center;
          justify-content: space-between;
          gap: var(--space-sm);
          margin-bottom: var(--space-xs);
        }

        .tile-header h3 {
          font-size: 0.9rem;
          font-weight: 600;
          color: var(--color-text);
          margin: 0;
          white-space: nowrap;
          overflow: hidden;
          text-overflow: ellipsis;
        }

        .tile-metric {
          font-family: var(--font-mono);
          font-size: 0.7rem;
          font-weight: 600;
          color: var(--tile-color);
          padding: 2px 6px;
          background: rgba(255, 255, 255, 0.05);
          border-radius: 4px;
          flex-shrink: 0;
        }

        .dash-tile p {
          font-size: 0.8rem;
          color: var(--color-text-muted);
          margin: 0;
          line-height: 1.4;
        }

        .comparison-table {
          max-width: 900px;
          margin: 0 auto var(--space-3xl);
          text-align: center;
        }

        .comparison-table h3 {
          margin-bottom: var(--space-lg);
          color: var(--color-text-muted);
          text-transform: uppercase;
          letter-spacing: 0.1em;
          font-size: 0.85rem;
        }

        .comparison-grid-table {
          width: 100%;
          border-collapse: collapse;
          border-radius: 8px;
          overflow: hidden;
          background: var(--color-bg-tertiary);
        }

        .comparison-grid-table th,
        .comparison-grid-table td {
          padding: var(--space-sm) var(--space-md);
          font-size: 0.8rem;
          border: 1px solid var(--color-bg-tertiary);
        }

        .comparison-grid-table th {
          background: var(--color-bg-secondary);
          font-weight: 600;
          text-align: center;
        }

        .comparison-grid-table td {
          background: var(--color-bg);
        }

        .comparison-grid-table td:first-child {
          text-align: left;
          color: var(--color-text-secondary);
        }

        .comparison-grid-table td:not(:first-child) {
          text-align: center;
          font-weight: 600;
        }

        .comparison-grid-table .yes {
          color: #10b981;
        }

        .comparison-grid-table .no {
          color: var(--color-text-muted);
        }

        .comparison-grid-table .partial {
          color: #f59e0b;
        }

        .cta-row {
          display: flex;
          justify-content: center;
        }

        @media (max-width: 1200px) {
          .dash-grid {
            grid-template-columns: repeat(2, 1fr);
          }
        }

        @media (max-width: 600px) {
          .dash-grid {
            grid-template-columns: 1fr;
          }

          .dash-tile {
            flex-direction: column;
            text-align: center;
          }

          .tile-header {
            justify-content: center;
          }

          .tile-icon {
            margin: 0 auto;
          }

          .comparison-table {
            overflow-x: auto;
          }

          .comparison-grid-table th,
          .comparison-grid-table td {
            padding: var(--space-xs) var(--space-sm);
            font-size: 0.7rem;
          }
        }
      `}</style>
    </section>
  );
}
