// @a0n/gnosis/forkjoin-app/client/components/Sections/EcosystemProductSection.tsx
import {
  Activity,
  ArrowRight,
  BookOpen,
  Brain,
  Command,
  Compass,
  Database,
  FileSearch,
  Github,
  GitBranch,
  Globe,
  Hammer,
  HardDrive,
  KeyRound,
  Lock,
  MessageSquare,
  Monitor,
  Orbit,
  Radio,
  RefreshCw,
  Settings,
  Shield,
  Sparkles,
  Wallet,
  Waves,
  Workflow,
  Zap,
  type LucideIcon,
} from '@a0n/aeon-icons';
import { getFoundationProduct, type FoundationProductId } from '../stack-model';

const productIconMap: Record<string, LucideIcon> = {
  activity: Activity,
  command: Command,
  compass: Compass,
  database: Database,
  'file-search': FileSearch,
  'git-branch': GitBranch,
  globe: Globe,
  hammer: Hammer,
  'hard-drive': HardDrive,
  'key-round': KeyRound,
  lock: Lock,
  'message-square': MessageSquare,
  monitor: Monitor,
  orbit: Orbit,
  radio: Radio,
  'refresh-cw': RefreshCw,
  settings: Settings,
  shield: Shield,
  sparkles: Sparkles,
  wallet: Wallet,
  waves: Waves,
  workflow: Workflow,
  brain: Brain,
  zap: Zap,
};

interface EcosystemProductSectionProps {
  productId: FoundationProductId;
}

/**
 * Renders the Ecosystem Product Section view.
 */
export function EcosystemProductSection({
  productId,
}: EcosystemProductSectionProps) {
  const product = getFoundationProduct(productId);
  const ProductIcon = productIconMap[product.icon];
  const relatedProductIds =
    (product.relatedProductIds as readonly FoundationProductId[] | undefined) ??
    [];
  const relatedProducts = relatedProductIds.map((relatedId) =>
    getFoundationProduct(relatedId)
  );

  return (
    <section
      id={product.id}
      className="section ecosystem-product-section"
      aria-labelledby={`${product.id}-title`}
      style={
        {
          '--product-color': product.color,
        } as React.CSSProperties
      }
    >
      <div className="container">
        <div className="section-header">
          <span className="layer-label">{product.layer}</span>
          <h2 id={`${product.id}-title`} className="product-section-title">
            {product.name}
          </h2>
          <p className="section-description">{product.tagline}</p>
        </div>

        <div className="ecosystem-product-grid">
          <div className="ecosystem-product-story">
            <div className="ecosystem-product-mark">
              <ProductIcon size={30} />
            </div>
            <p>{product.description}</p>
            <div className="ecosystem-product-actions">
              <a href={product.href} className="button button-primary">
                <ArrowRight size={18} />
                Explore {product.name}
              </a>
              {product.docsUrl ? (
                <a
                  href={product.docsUrl}
                  className="button button-secondary"
                  target="_blank"
                  rel="noopener noreferrer"
                >
                  <BookOpen size={18} />
                  Docs
                </a>
              ) : null}
              {product.sourceUrl ? (
                <a
                  href={product.sourceUrl}
                  className="button button-secondary"
                  target="_blank"
                  rel="noopener noreferrer"
                >
                  <Github size={18} />
                  Source
                </a>
              ) : null}
            </div>
          </div>

          <div className="ecosystem-product-panels">
            <div className="ecosystem-product-panel">
              <h3>Core Capabilities</h3>
              <div className="ecosystem-chip-grid">
                {product.highlights.map((highlight) => (
                  <span key={highlight} className="ecosystem-chip">
                    {highlight}
                  </span>
                ))}
              </div>
            </div>

            {relatedProducts.length > 0 ? (
              <div className="ecosystem-product-panel">
                <h3>Connected Layers</h3>
                <div className="ecosystem-related-grid">
                  {relatedProducts.map((relatedProduct) => (
                    <a
                      key={relatedProduct.id}
                      href={relatedProduct.href}
                      className="ecosystem-related-card"
                    >
                      <span className="ecosystem-related-name">
                        {relatedProduct.name}
                      </span>
                      <span className="ecosystem-related-layer">
                        {relatedProduct.layer}
                      </span>
                    </a>
                  ))}
                </div>
              </div>
            ) : null}
          </div>
        </div>
      </div>

      <style>{`
        .ecosystem-product-section {
          background:
            radial-gradient(circle at top left, color-mix(in srgb, var(--product-color) 16%, transparent) 0%, transparent 48%),
            linear-gradient(180deg, var(--color-bg) 0%, var(--color-bg-secondary) 100%);
        }

        .ecosystem-product-grid {
          display: grid;
          grid-template-columns: minmax(0, 1.1fr) minmax(0, 1fr);
          gap: var(--space-xl);
          align-items: start;
        }

        .ecosystem-product-story,
        .ecosystem-product-panel {
          background: rgba(255, 255, 255, 0.03);
          border: 1px solid rgba(255, 255, 255, 0.08);
          border-radius: 20px;
          padding: var(--space-xl);
          backdrop-filter: blur(10px);
        }

        .ecosystem-product-story {
          min-height: 100%;
        }

        .ecosystem-product-mark {
          width: 64px;
          height: 64px;
          display: inline-flex;
          align-items: center;
          justify-content: center;
          border-radius: 18px;
          color: var(--product-color);
          background: color-mix(in srgb, var(--product-color) 18%, transparent);
          margin-bottom: var(--space-lg);
        }

        .ecosystem-product-story p {
          margin: 0 0 var(--space-lg);
          color: var(--color-text-secondary);
          line-height: 1.8;
        }

        .ecosystem-product-actions {
          display: flex;
          gap: var(--space-sm);
          flex-wrap: wrap;
        }

        .ecosystem-product-actions .button {
          display: inline-flex;
          align-items: center;
          gap: var(--space-xs);
        }

        .ecosystem-product-panels {
          display: flex;
          flex-direction: column;
          gap: var(--space-lg);
        }

        .ecosystem-product-panel h3 {
          margin-bottom: var(--space-md);
        }

        .ecosystem-chip-grid {
          display: flex;
          flex-wrap: wrap;
          gap: 10px;
        }

        .ecosystem-chip {
          padding: 0.55rem 0.85rem;
          border-radius: 999px;
          background: color-mix(in srgb, var(--product-color) 16%, rgba(255, 255, 255, 0.04));
          border: 1px solid color-mix(in srgb, var(--product-color) 40%, rgba(255, 255, 255, 0.1));
          color: var(--color-text);
          font-size: 0.88rem;
          line-height: 1.4;
        }

        .ecosystem-related-grid {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(180px, 1fr));
          gap: var(--space-sm);
        }

        .ecosystem-related-card {
          display: flex;
          flex-direction: column;
          gap: 0.35rem;
          padding: 0.9rem 1rem;
          border-radius: 16px;
          text-decoration: none;
          color: inherit;
          border: 1px solid rgba(255, 255, 255, 0.08);
          background: rgba(0, 0, 0, 0.18);
          transition: transform 0.2s ease, border-color 0.2s ease;
        }

        .ecosystem-related-card:hover {
          transform: translateY(-2px);
          border-color: color-mix(in srgb, var(--product-color) 52%, rgba(255, 255, 255, 0.12));
        }

        .ecosystem-related-name {
          font-weight: 600;
        }

        .ecosystem-related-layer {
          color: var(--color-text-muted);
          font-size: 0.8rem;
          text-transform: uppercase;
          letter-spacing: 0.08em;
        }

        @media (max-width: 960px) {
          .ecosystem-product-grid {
            grid-template-columns: 1fr;
          }
        }
      `}</style>
    </section>
  );
}
