// @a0n/gnosis/forkjoin-app/client/components/Sections/EdgeworkSection.tsx
/**
 * EdgeworkSection - Edge Compute & On-Device AI
 *
 * Browser-first AI inference and storage from edgework-sdk:
 * - WebGPU-powered inference (on-device AI)
 * - Local-first storage with D1/Dash sync
 * - RLHF training (improve models over time)
 * - React components for feedback UI
 */

import {
  Globe,
  Cpu,
  Cloud,
  Server,
  Zap,
  Timer,
  Shield,
  Brain,
  Database,
  ThumbsUp,
  Box,
  Layers,
  WifiOff,
  Gauge,
  Sparkles,
  Code,
} from '@a0n/aeon-icons';

const features = [
  // Row 1: Inference
  {
    icon: Brain,
    title: 'WebGPU Inference',
    metric: 'On-device',
    description:
      'AI runs entirely in browser via WebGPU. Fast, private, no server calls.',
    category: 'inference',
  },
  {
    icon: Cpu,
    title: 'WebNN Support',
    metric: 'NPU',
    description:
      'Apple Neural Engine, Qualcomm Hexagon. Hardware AI acceleration.',
    category: 'inference',
  },
  {
    icon: Sparkles,
    title: 'Text Generation',
    metric: 'LLM',
    description:
      'Run language models like Cyrano-360M directly in browser. Streaming output.',
    category: 'inference',
  },
  {
    icon: Gauge,
    title: 'WASM Fallback',
    metric: 'Universal',
    description:
      'WASM backend when WebGPU unavailable. Works on any modern browser.',
    category: 'inference',
  },

  // Row 2: Storage
  {
    icon: Database,
    title: 'Local-First Data',
    metric: 'IndexedDB',
    description:
      'Data lives in browser. IndexedDB, OPFS, memory adapters included.',
    category: 'storage',
  },
  {
    icon: Layers,
    title: 'D1/Dash Sync',
    metric: 'Cloud',
    description:
      'Sync to Cloudflare D1 and Dash. Local-first with cloud backup.',
    category: 'storage',
  },
  {
    icon: WifiOff,
    title: 'Offline-First',
    metric: 'Queue',
    description: 'Works offline. Queue operations, sync when back online.',
    category: 'storage',
  },
  {
    icon: Box,
    title: 'Model Caching',
    metric: 'OPFS',
    description: 'Cache model weights in OPFS. Download once, run forever.',
    category: 'storage',
  },

  // Row 3: RLHF & Feedback
  {
    icon: ThumbsUp,
    title: 'RLHF Training',
    metric: 'On-device',
    description:
      'Collect feedback to improve models. Learning happens locally.',
    category: 'rlhf',
  },
  {
    icon: Code,
    title: 'React Components',
    metric: 'Ready',
    description:
      'EdgeworkProvider, useInference hooks. Drop-in React integration.',
    category: 'rlhf',
  },
  {
    icon: Shield,
    title: 'Privacy First',
    metric: 'Local',
    description: 'Your data never leaves the device. AI runs where you are.',
    category: 'rlhf',
  },
  {
    icon: Timer,
    title: 'Batch Collection',
    metric: 'Async',
    description:
      'Batch feedback uploads. Efficient, non-blocking, background sync.',
    category: 'rlhf',
  },

  // Row 4: Compute Pipeline
  {
    icon: Globe,
    title: 'Browser First',
    metric: '0ms',
    description:
      'Start in browser instantly. No server round-trip for first inference.',
    category: 'compute',
  },
  {
    icon: Zap,
    title: 'Edge Fallback',
    metric: '5ms',
    description:
      "Fall back to edge when browser can't handle it. Cloudflare Workers AI.",
    category: 'compute',
  },
  {
    icon: Server,
    title: 'Cloud Backup',
    metric: '50ms',
    description:
      'Heavy workloads go to cloud. Seamless escalation when needed.',
    category: 'compute',
  },
  {
    icon: Cloud,
    title: 'Unified SDK',
    metric: 'One API',
    description: 'Same API for browser, edge, cloud. Write once, run anywhere.',
    category: 'compute',
  },
];

const categoryColors: Record<string, string> = {
  inference: '#3b82f6',
  storage: '#f59e0b',
  rlhf: '#10b981',
  compute: '#06b6d4',
};

/**
 * Renders the Edgework Section view.
 */
export function EdgeworkSection() {
  return (
    <section
      id="edgework"
      className="section edgework-section"
      aria-labelledby="edgework-title"
    >
      <div className="container">
        <div className="section-header">
          <span className="layer-label">Compute Layer</span>
          <h2 id="edgework-title" className="product-section-title">
            Edgework
          </h2>
          <p className="section-description">
            Browser to edge to cloud in milliseconds.
          </p>
        </div>

        <div className="pipeline-visual">
          <div className="pipeline-stage">
            <Globe size={28} />
            <h3>Browser</h3>
            <span className="stage-time">0ms</span>
          </div>
          <span className="arrow" aria-hidden="true">
            &rarr;
          </span>
          <div className="pipeline-stage">
            <Cpu size={28} />
            <h3>Edge</h3>
            <span className="stage-time">5ms</span>
          </div>
          <span className="arrow" aria-hidden="true">
            &rarr;
          </span>
          <div className="pipeline-stage">
            <Server size={28} />
            <h3>Infra</h3>
            <span className="stage-time">10ms</span>
          </div>
          <span className="arrow" aria-hidden="true">
            &rarr;
          </span>
          <div className="pipeline-stage">
            <Cloud size={28} />
            <h3>Cloud</h3>
            <span className="stage-time">50ms</span>
          </div>
        </div>

        <div className="edgework-grid">
          {features.map((feature) => (
            <div
              key={feature.title}
              className="edgework-tile"
              style={
                {
                  '--tile-color': categoryColors[feature.category],
                } as React.CSSProperties
              }
            >
              <feature.icon size={24} className="tile-icon" />
              <div className="tile-content">
                <div className="tile-header">
                  <h3>{feature.title}</h3>
                  <span className="tile-metric">{feature.metric}</span>
                </div>
                <p>{feature.description}</p>
              </div>
            </div>
          ))}
        </div>

        <div className="comparison-table">
          <h3>vs Other AI Runtimes</h3>
          <table className="comparison-grid-table">
            <thead>
              <tr>
                <th>Feature</th>
                <th>Edgework</th>
                <th>Transformers.js</th>
                <th>TensorFlow.js</th>
                <th>ONNX Web</th>
                <th>Workers AI</th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>WebGPU Inference</td>
                <td className="yes">✓ Native</td>
                <td className="yes">✓ Yes</td>
                <td className="partial">Partial</td>
                <td className="yes">✓ Yes</td>
                <td className="no">No</td>
              </tr>
              <tr>
                <td>WebNN Support</td>
                <td className="yes">✓ Yes</td>
                <td className="partial">Experimental</td>
                <td className="no">No</td>
                <td className="yes">✓ Yes</td>
                <td className="no">No</td>
              </tr>
              <tr>
                <td>Local-First Storage</td>
                <td className="yes">✓ Built-in</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
              </tr>
              <tr>
                <td>RLHF/Feedback</td>
                <td className="yes">✓ Built-in</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
              </tr>
              <tr>
                <td>Model Caching (OPFS)</td>
                <td className="yes">✓ Yes</td>
                <td className="yes">✓ Yes</td>
                <td className="partial">Basic</td>
                <td className="partial">Basic</td>
                <td className="no">N/A</td>
              </tr>
              <tr>
                <td>React Components</td>
                <td className="yes">✓ Native</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
              </tr>
              <tr>
                <td>Edge Fallback</td>
                <td className="yes">✓ Automatic</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="yes">✓ Native</td>
              </tr>
              <tr>
                <td>Unified SDK</td>
                <td className="yes">✓ One API</td>
                <td className="no">Browser only</td>
                <td className="partial">Partial</td>
                <td className="no">Browser only</td>
                <td className="no">Edge only</td>
              </tr>
            </tbody>
          </table>
        </div>

        <div className="cta-row">
          <a href="#/edgework" className="button button-primary">
            Explore Edgework
          </a>
        </div>
      </div>

      <style>{`
        .edgework-section {
          --ew-blue: #3b82f6;
          background: #000000;
          color: #f8fafc;
        }

        .section-header {
          text-align: center;
          margin-bottom: var(--space-3xl);
        }

        .section-description {
          font-size: 1.125rem;
          color: rgba(248, 250, 252, 0.7);
          max-width: 600px;
          margin: var(--space-md) auto 0;
        }

        .pipeline-visual {
          display: flex;
          align-items: center;
          justify-content: center;
          gap: var(--space-md);
          margin-bottom: var(--space-3xl);
          flex-wrap: wrap;
        }

        .pipeline-stage {
          text-align: center;
          padding: var(--space-lg);
          background: rgba(59, 130, 246, 0.05);
          border: 1px solid rgba(59, 130, 246, 0.2);
          border-radius: 12px;
          min-width: 100px;
        }

        .pipeline-stage svg {
          color: var(--ew-blue);
          margin-bottom: var(--space-xs);
        }

        .pipeline-stage h3 {
          font-size: 0.9rem;
          margin-bottom: var(--space-xs);
        }

        .stage-time {
          font-family: var(--font-mono);
          font-size: 0.8rem;
          color: #60a5fa;
        }

        .arrow {
          color: rgba(59, 130, 246, 0.5);
          font-size: 1.2rem;
        }

        .edgework-grid {
          display: grid;
          grid-template-columns: repeat(4, 1fr);
          gap: var(--space-md);
          margin-bottom: var(--space-3xl);
        }

        .edgework-tile {
          display: flex;
          gap: var(--space-md);
          padding: var(--space-md);
          background: rgba(255, 255, 255, 0.02);
          border: 1px solid rgba(255, 255, 255, 0.1);
          border-radius: 10px;
          transition: all 0.2s ease;
        }

        .edgework-tile:hover {
          border-color: var(--tile-color);
          transform: translateY(-2px);
        }

        .tile-icon {
          color: var(--tile-color);
          flex-shrink: 0;
          margin-top: 2px;
        }

        .tile-content {
          flex: 1;
          min-width: 0;
        }

        .tile-header {
          display: flex;
          align-items: center;
          justify-content: space-between;
          gap: var(--space-sm);
          margin-bottom: var(--space-xs);
        }

        .tile-header h3 {
          font-size: 0.9rem;
          font-weight: 600;
          color: #f8fafc;
          margin: 0;
          white-space: nowrap;
          overflow: hidden;
          text-overflow: ellipsis;
        }

        .tile-metric {
          font-family: var(--font-mono);
          font-size: 0.7rem;
          font-weight: 600;
          color: var(--tile-color);
          padding: 2px 6px;
          background: rgba(255, 255, 255, 0.05);
          border-radius: 4px;
          flex-shrink: 0;
        }

        .edgework-tile p {
          font-size: 0.8rem;
          color: rgba(248, 250, 252, 0.75);
          margin: 0;
          line-height: 1.4;
        }

        .comparison-table {
          max-width: 900px;
          margin: 0 auto var(--space-3xl);
          text-align: center;
        }

        .comparison-table h3 {
          margin-bottom: var(--space-lg);
          color: rgba(248, 250, 252, 0.7);
          text-transform: uppercase;
          letter-spacing: 0.1em;
          font-size: 0.85rem;
        }

        .comparison-grid-table {
          width: 100%;
          border-collapse: collapse;
          border-radius: 8px;
          overflow: hidden;
          background: rgba(255, 255, 255, 0.02);
        }

        .comparison-grid-table th,
        .comparison-grid-table td {
          padding: var(--space-sm) var(--space-md);
          font-size: 0.8rem;
          border: 1px solid rgba(255, 255, 255, 0.1);
        }

        .comparison-grid-table th {
          background: rgba(255, 255, 255, 0.05);
          font-weight: 600;
          text-align: center;
          color: #f8fafc;
        }

        .comparison-grid-table td {
          background: rgba(0, 0, 0, 0.3);
        }

        .comparison-grid-table td:first-child {
          text-align: left;
          color: rgba(248, 250, 252, 0.7);
        }

        .comparison-grid-table td:not(:first-child) {
          text-align: center;
          font-weight: 600;
        }

        .comparison-grid-table .yes {
          color: #10b981;
        }

        .comparison-grid-table .no {
          color: rgba(248, 250, 252, 0.7);
        }

        .comparison-grid-table .partial {
          color: #f59e0b;
        }

        .cta-row {
          display: flex;
          justify-content: center;
        }

        @media (max-width: 1200px) {
          .edgework-grid {
            grid-template-columns: repeat(2, 1fr);
          }
        }

        @media (max-width: 768px) {
          .pipeline-visual {
            flex-direction: column;
          }
          .arrow {
            transform: rotate(90deg);
          }
        }

        @media (max-width: 600px) {
          .edgework-grid {
            grid-template-columns: 1fr;
          }

          .edgework-tile {
            flex-direction: column;
            text-align: center;
          }

          .tile-header {
            justify-content: center;
          }

          .tile-icon {
            margin: 0 auto;
          }

          .comparison-table {
            overflow-x: auto;
          }

          .comparison-grid-table th,
          .comparison-grid-table td {
            padding: var(--space-xs) var(--space-sm);
            font-size: 0.7rem;
          }
        }
      `}</style>
    </section>
  );
}
