// @a0n/gnosis/forkjoin-app/client/components/Sections/FluxSection.tsx
/**
 * FluxSection - The CMS IS the Website
 *
 * Shows the incredible capabilities from the Aeon-Flux README:
 * - Zero-Dependency Rendering (WASM, inline everything)
 * - Hyperpersonalized Routing (routes adapt to user)
 * - Edge Side Inference (ESI - AI at render time)
 * - Visual Editing to GitHub PR publishing
 * - Speculative Pre-rendering (instant navigation)
 */

import {
  Globe,
  Brain,
  GitBranch,
  Zap,
  WifiOff,
  FileCode,
  Eye,
  Layers,
  Timer,
  MousePointer,
  Sparkles,
  PenTool,
  GitPullRequest,
  Box,
  Gauge,
  Database,
  MousePointerClick,
  Binary,
} from '@a0n/aeon-icons';

const features = [
  // Row 1: Edge Rendering
  {
    icon: Database,
    title: 'Database First',
    metric: 'No files',
    description:
      'Pages live in the database. No file system, no build step, no waiting.',
    category: 'rendering',
  },
  {
    icon: Globe,
    title: 'Edge Native',
    metric: '300+ PoPs',
    description:
      'Renders at the edge, close to users. Cloudflare Workers powered.',
    category: 'rendering',
  },
  {
    icon: FileCode,
    title: 'Inlined CSS',
    metric: 'Critical',
    description:
      'CSS inlined in HTML. No render-blocking requests. Instant style.',
    category: 'rendering',
  },
  {
    icon: Gauge,
    title: 'WASM Runtime',
    metric: '~180KB',
    description:
      'Lightweight WASM runtime vs 500KB+ Next.js. Instant loading on any connection.',
    category: 'rendering',
  },

  // Row 2: Performance
  {
    icon: Layers,
    title: 'Zero CLS',
    metric: '0',
    description:
      'No layout shift. Content sized at render time. Perfect Core Web Vitals.',
    category: 'performance',
  },
  {
    icon: Box,
    title: 'Free Skeleton',
    metric: 'Auto',
    description:
      'Skeleton states generated automatically. No extra code. Instant feedback.',
    category: 'performance',
  },
  {
    icon: Eye,
    title: 'Lazy Hydration',
    metric: 'Visibility',
    description:
      'Only hydrate components when visible. IntersectionObserver-powered.',
    category: 'performance',
  },
  {
    icon: Timer,
    title: 'Multi-Layer Cache',
    metric: 'KV→D1',
    description:
      'KV (1ms) to D1 (5ms) to Session (50ms). Every request is a cache hit.',
    category: 'performance',
  },

  // Row 3: Navigation + ESI
  {
    icon: MousePointer,
    title: 'Speculative Render',
    metric: '0ms nav',
    description:
      'Pages render before click. Hover prediction. Click = instant navigation.',
    category: 'routing',
  },
  {
    icon: Sparkles,
    title: 'Hyperpersonalized',
    metric: '0.014ms',
    description:
      'Routes adapt to user: time, location, emotion, tier. The website comes to you.',
    category: 'routing',
  },
  {
    icon: Brain,
    title: 'ESI.Infer',
    metric: 'LLM',
    description:
      'AI inference at render time. Personalize content before reaching browser.',
    category: 'esi',
  },
  {
    icon: Zap,
    title: 'Structured Output',
    metric: 'Zod',
    description:
      'Type-safe AI responses with Zod schemas. Guaranteed structure.',
    category: 'esi',
  },

  // Row 4: CMS & Publishing
  {
    icon: PenTool,
    title: 'Visual Editing',
    metric: 'In-place',
    description:
      'Click to edit in the browser. No separate admin panel. CMS IS the site.',
    category: 'cms',
  },
  {
    icon: GitPullRequest,
    title: 'GitHub Publishing',
    metric: 'Auto PR',
    description:
      'Visual edits generate TSX, create PR, CI deploys. One-click publish.',
    category: 'cms',
  },
  {
    icon: GitBranch,
    title: "'use aeon'",
    metric: 'Directive',
    description:
      'One directive enables everything. Presence, sync, version, data hooks.',
    category: 'cms',
  },
  {
    icon: WifiOff,
    title: 'Offline-First',
    metric: 'Queue',
    description:
      'Queue edits offline. Sync when back. Encrypted at rest. Works anywhere.',
    category: 'cms',
  },

  // Row 5: Analytics
  {
    icon: MousePointerClick,
    title: 'Click Tracking',
    metric: 'Zero-code',
    description:
      'Every click tracked automatically. No instrumentation needed. Full context.',
    category: 'analytics',
  },
  {
    icon: Binary,
    title: 'Merkle Tree IDs',
    metric: 'Stable',
    description:
      'Content-addressable node hashes. Same content = same ID across renders.',
    category: 'analytics',
  },
  {
    icon: Layers,
    title: 'Tree Paths',
    metric: 'Full ancestry',
    description:
      'Every click includes full tree path from root. Know exactly what was clicked.',
    category: 'analytics',
  },
  {
    icon: Brain,
    title: 'ESI Context',
    metric: 'Rich',
    description:
      'User tier, emotion, features, session. Complete context with every event.',
    category: 'analytics',
  },
];

const categoryColors: Record<string, string> = {
  rendering: '#10b981',
  performance: '#3b82f6',
  routing: '#06b6d4',
  esi: '#5599cc',
  cms: '#f59e0b',
  analytics: '#a855f7',
};

/**
 * Renders the Flux Section view.
 */
export function FluxSection() {
  return (
    <section
      id="flux"
      className="section flux-section"
      aria-labelledby="flux-title"
    >
      <div className="container">
        <div className="section-header">
          <span className="layer-label">Surface Layer</span>
          <h2 id="flux-title" className="product-section-title">
            Aeon-Flux
          </h2>
          <p className="section-description">
            Edge-first rendering. AI at render time. Instant navigation.
          </p>
        </div>

        <div className="metrics-row">
          <div className="metric-card">
            <span className="metric-value">180KB</span>
            <span className="metric-label">Runtime size</span>
            <span className="metric-comparison">vs 500KB+</span>
          </div>
          <div className="metric-card">
            <span className="metric-value">7ms</span>
            <span className="metric-label">Edge render</span>
            <span className="metric-comparison">vs 50ms+</span>
          </div>
          <div className="metric-card">
            <span className="metric-value">0ms</span>
            <span className="metric-label">Navigation</span>
            <span className="metric-comparison">Speculative</span>
          </div>
        </div>

        <div className="manifesto">
          <h3>
            The CMS <em>is</em> the website.
          </h3>
          <p>
            Edit visually in the browser. Changes generate TSX. A PR is created.
            CI deploys. No developer context-switching. No separate admin panel.
            The website and the CMS are one.
          </p>
        </div>

        <div className="flux-grid">
          {features.map((feature) => (
            <div
              key={feature.title}
              className="flux-tile"
              style={
                {
                  '--tile-color': categoryColors[feature.category],
                } as React.CSSProperties
              }
            >
              <feature.icon size={24} className="tile-icon" />
              <div className="tile-content">
                <div className="tile-header">
                  <h3>{feature.title}</h3>
                  <span className="tile-metric">{feature.metric}</span>
                </div>
                <p>{feature.description}</p>
              </div>
            </div>
          ))}
        </div>

        <div className="comparison-table">
          <h3>vs Next.js</h3>
          <div className="comparison-grid">
            <div className="comparison-header">Metric</div>
            <div className="comparison-header">Next.js</div>
            <div className="comparison-header">Aeon-Flux</div>

            <div className="comparison-feature">Runtime</div>
            <div className="comparison-value old">~500KB</div>
            <div className="comparison-value new">~180KB</div>
            <div className="comparison-feature">Requests</div>
            <div className="comparison-value old">15-30</div>
            <div className="comparison-value new">1</div>

            <div className="comparison-feature">TTFB</div>
            <div className="comparison-value old">100ms</div>
            <div className="comparison-value new">50ms</div>

            <div className="comparison-feature">First Paint</div>
            <div className="comparison-value old">500ms</div>
            <div className="comparison-value new">&lt;100ms</div>

            <div className="comparison-feature">AI Built-in</div>
            <div className="comparison-value old">Add-on</div>
            <div className="comparison-value new">ESI</div>

            <div className="comparison-feature">CMS</div>
            <div className="comparison-value old">Separate</div>
            <div className="comparison-value new">IS the site</div>

            <div className="comparison-feature">Analytics</div>
            <div className="comparison-value old">Manual</div>
            <div className="comparison-value new">Zero-code</div>
          </div>
        </div>

        <div className="cta-row">
          <a href="#/flux" className="button button-primary">
            Explore Aeon-Flux
          </a>
        </div>
      </div>

      <style>{`
        .flux-section {
          --flux-emerald: #10b981;
          background: var(--color-bg-secondary);
          border-bottom: 1px solid var(--color-bg-tertiary);
        }

        .section-header {
          text-align: center;
          margin-bottom: var(--space-3xl);
        }

        .section-description {
          font-size: 1.125rem;
          color: var(--color-text-secondary);
          max-width: 600px;
          margin: var(--space-md) auto 0;
        }

        .metrics-row {
          display: flex;
          justify-content: center;
          gap: var(--space-xl);
          flex-wrap: wrap;
          margin-bottom: var(--space-3xl);
        }

        .metric-card {
          text-align: center;
          padding: var(--space-lg);
        }

        .metric-value {
          display: block;
          font-size: 3rem;
          font-weight: 700;
          color: var(--flux-emerald);
          line-height: 1;
        }

        .metric-label {
          display: block;
          font-size: 1rem;
          color: var(--color-text);
          margin-top: var(--space-sm);
        }

        .metric-comparison {
          display: block;
          font-size: 0.75rem;
          color: var(--color-text-muted);
          margin-top: var(--space-xs);
        }

        .manifesto {
          text-align: center;
          margin-bottom: var(--space-3xl);
          padding: var(--space-xl);
          background: var(--color-bg);
          border-radius: var(--radius-lg);
          border: 1px solid var(--color-bg-tertiary);
        }

        .manifesto h3 {
          font-size: 1.75rem;
          margin-bottom: var(--space-md);
        }

        .manifesto em {
          color: var(--flux-emerald);
          font-style: normal;
        }

        .manifesto p {
          max-width: 600px;
          margin: 0 auto;
          color: var(--color-text-secondary);
        }

        .flux-grid {
          display: grid;
          grid-template-columns: repeat(4, 1fr);
          gap: var(--space-md);
          margin-bottom: var(--space-3xl);
        }

        .flux-tile {
          display: flex;
          gap: var(--space-md);
          padding: var(--space-md);
          background: var(--color-bg);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 10px;
          transition: all 0.2s ease;
        }

        .flux-tile:hover {
          border-color: var(--tile-color);
          transform: translateY(-2px);
        }

        .tile-icon {
          color: var(--tile-color);
          flex-shrink: 0;
          margin-top: 2px;
        }

        .tile-content {
          flex: 1;
          min-width: 0;
        }

        .tile-header {
          display: flex;
          align-items: center;
          justify-content: space-between;
          gap: var(--space-sm);
          margin-bottom: var(--space-xs);
        }

        .tile-header h3 {
          font-size: 0.9rem;
          font-weight: 600;
          color: var(--color-text);
          margin: 0;
          white-space: nowrap;
          overflow: hidden;
          text-overflow: ellipsis;
        }

        .tile-metric {
          font-family: var(--font-mono);
          font-size: 0.7rem;
          font-weight: 600;
          color: var(--tile-color);
          padding: 2px 6px;
          background: rgba(255, 255, 255, 0.05);
          border-radius: 4px;
          flex-shrink: 0;
        }

        .flux-tile p {
          font-size: 0.8rem;
          color: var(--color-text-muted);
          margin: 0;
          line-height: 1.4;
        }

        .comparison-table {
          max-width: 500px;
          margin: 0 auto var(--space-3xl);
          text-align: center;
        }

        .comparison-table h3 {
          margin-bottom: var(--space-lg);
          color: var(--color-text-muted);
          text-transform: uppercase;
          letter-spacing: 0.1em;
          font-size: 0.85rem;
        }

        .comparison-grid {
          display: grid;
          grid-template-columns: 1.5fr 1fr 1fr;
          gap: 1px;
          background: var(--color-bg-tertiary);
          border-radius: 8px;
          overflow: hidden;
        }

        .comparison-header {
          background: var(--color-bg);
          padding: var(--space-sm) var(--space-md);
          font-weight: 600;
          font-size: 0.85rem;
        }

        .comparison-feature {
          background: var(--color-bg);
          padding: var(--space-sm) var(--space-md);
          text-align: left;
          font-size: 0.85rem;
          color: var(--color-text-secondary);
        }

        .comparison-value {
          background: var(--color-bg);
          padding: var(--space-sm) var(--space-md);
          font-size: 0.85rem;
          font-weight: 600;
        }

        .comparison-value.old {
          color: var(--color-text-muted);
        }

        .comparison-value.new {
          color: var(--flux-emerald);
        }

        .cta-row {
          display: flex;
          justify-content: center;
        }

        @media (max-width: 1200px) {
          .flux-grid {
            grid-template-columns: repeat(2, 1fr);
          }
        }

        @media (max-width: 600px) {
          .flux-grid {
            grid-template-columns: 1fr;
          }

          .flux-tile {
            flex-direction: column;
            text-align: center;
          }

          .tile-header {
            justify-content: center;
          }

          .tile-icon {
            margin: 0 auto;
          }
        }
      `}</style>
    </section>
  );
}
