// @a0n/gnosis/forkjoin-app/client/components/Sections/GetStartedSection.tsx
import { memo, useState } from '@a0n/raect';
import {
  Copy,
  Check,
  BookOpen,
  Star,
  MessageCircle,
  FlaskConical,
  Brain,
} from '@a0n/aeon-icons';
import { FOUNDATION_DOCS_QUICKSTART_URL } from '../../docsLinks';

export const GetStartedSection = memo(function GetStartedSection() {
  const [copied, setCopied] = useState<string | null>(null);

  const copyToClipboard = (text: string, id: string) => {
    navigator.clipboard.writeText(text);
    setCopied(id);
    setTimeout(() => setCopied(null), 2000);
  };

  return (
    <section id="get-started" className="section get-started-section">
      <div className="container">
        <div className="section-header">
          <h2>Get Started</h2>
          <p className="section-subtitle">
            Escape in 30 seconds. Migrate in 5 minutes.
          </p>
        </div>

        <div className="options-grid">
          <div className="option-card primary">
            <h3>New Project</h3>
            <p>Start fresh with everything configured.</p>
            <div className="command-block">
              <code>npx create-aeon-flux my-app</code>
              <button
                className="copy-btn"
                onClick={() =>
                  copyToClipboard('npx create-aeon-flux my-app', 'new')
                }
                aria-label="Copy command"
              >
                {copied === 'new' ? <Check size={16} /> : <Copy size={16} />}
              </button>
            </div>
            <ul className="feature-list">
              <li>Edge-first project structure</li>
              <li>Cloudflare Workers preconfigured</li>
              <li>ESI and presence ready</li>
              <li>TypeScript, ESLint, Prettier</li>
            </ul>
          </div>

          <div className="option-card">
            <h3>Migrate Existing</h3>
            <p>
              Convert your existing React app. Keep your code, lose the bloat.
            </p>
            <div className="command-block">
              <code>npx aeon-flux migrate ./my-app</code>
              <button
                className="copy-btn"
                onClick={() =>
                  copyToClipboard('npx aeon-flux migrate ./my-app', 'migrate')
                }
                aria-label="Copy command"
              >
                {copied === 'migrate' ? (
                  <Check size={16} />
                ) : (
                  <Copy size={16} />
                )}
              </button>
            </div>
            <ul className="feature-list">
              <li>Auto-detects your framework</li>
              <li>Converts routing automatically</li>
              <li>Preserves your components</li>
              <li>Generates wrangler.toml</li>
            </ul>
          </div>

          <div className="option-card">
            <h3>Just the Sync</h3>
            <p>Add Aeon to any project for distributed state.</p>
            <div className="command-block">
              <code>npm install @a0n/aeon</code>
              <button
                className="copy-btn"
                onClick={() => copyToClipboard('npm install @a0n/aeon', 'aeon')}
                aria-label="Copy command"
              >
                {copied === 'aeon' ? <Check size={16} /> : <Copy size={16} />}
              </button>
            </div>
            <ul className="feature-list">
              <li>Framework agnostic</li>
              <li>Works with any backend</li>
              <li>Add presence to existing apps</li>
              <li>Offline-first sync</li>
            </ul>
          </div>
        </div>

        <div className="resources">
          <h3>Resources</h3>
          <div className="resources-grid">
            <a
              href={FOUNDATION_DOCS_QUICKSTART_URL}
              className="resource-link"
              target="_blank"
              rel="noopener noreferrer"
            >
              <span className="resource-icon">
                <BookOpen size={24} />
              </span>
              <span className="resource-text">
                <strong>Documentation</strong>
                <span>Guides, API reference, examples</span>
              </span>
            </a>
            <a
              href="https://github.com/forkjoin-ai/aeon"
              className="resource-link"
              target="_blank"
              rel="noopener noreferrer"
            >
              <span className="resource-icon">
                <Star size={24} />
              </span>
              <span className="resource-text">
                <strong>GitHub</strong>
                <span>Star us, contribute, report issues</span>
              </span>
            </a>
            <a
              href="https://discord.gg/aeon"
              className="resource-link"
              target="_blank"
              rel="noopener noreferrer"
            >
              <span className="resource-icon">
                <MessageCircle size={24} />
              </span>
              <span className="resource-text">
                <strong>Discord</strong>
                <span>Join the community</span>
              </span>
            </a>
            <a
              href="https://github.com/forkjoin-ai/aeon/tree/main/examples"
              className="resource-link"
              target="_blank"
              rel="noopener noreferrer"
            >
              <span className="resource-icon">
                <FlaskConical size={24} />
              </span>
              <span className="resource-text">
                <strong>Examples</strong>
                <span>Real-world implementations</span>
              </span>
            </a>
            <a
              href="https://huggingface.co/affectively"
              className="resource-link"
              target="_blank"
              rel="noopener noreferrer"
            >
              <span className="resource-icon">
                <Brain size={24} />
              </span>
              <span className="resource-text">
                <strong>HuggingFace</strong>
                <span>Pre-trained models</span>
              </span>
            </a>
          </div>
        </div>
      </div>

      <style>{`
        .get-started-section {
          background: var(--color-bg-secondary);
        }

        .section-header {
          text-align: center;
          margin-bottom: var(--space-3xl);
        }

        .section-subtitle {
          font-size: 1.25rem;
          max-width: 600px;
          margin: 0 auto;
        }

        .options-grid {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(300px, 1fr));
          gap: var(--space-xl);
          margin-bottom: var(--space-3xl);
        }

        .option-card {
          background: var(--color-bg);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: var(--radius-lg);
          padding: var(--space-xl);
          transition: all var(--transition-normal);
        }

        .option-card:hover {
          border-color: var(--color-accent);
          box-shadow: 0 0 30px var(--color-accent-glow);
        }

        .option-card.primary {
          border-color: var(--color-accent);
          background: linear-gradient(135deg, var(--color-bg) 0%, rgba(99, 102, 241, 0.05) 100%);
        }

        .option-card h3 {
          margin-bottom: var(--space-sm);
        }

        .option-card > p {
          font-size: 0.875rem;
          margin-bottom: var(--space-lg);
        }

        .command-block {
          display: flex;
          align-items: center;
          gap: var(--space-sm);
          padding: var(--space-md);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: var(--radius-md);
          margin-bottom: var(--space-lg);
        }

        .command-block code {
          flex: 1;
          font-family: var(--font-mono);
          font-size: 0.875rem;
          color: var(--color-accent);
          background: none;
          padding: 0;
        }

        .copy-btn {
          background: none;
          border: none;
          cursor: pointer;
          font-size: 1rem;
          padding: var(--space-xs);
          opacity: 0.6;
          transition: opacity var(--transition-fast);
        }

        .copy-btn:hover {
          opacity: 1;
        }

        .feature-list {
          list-style: none;
          padding: 0;
          margin: 0;
        }

        .feature-list li {
          font-size: 0.875rem;
          color: var(--color-text-secondary);
          padding: var(--space-xs) 0;
          padding-left: 1.5rem;
          position: relative;
        }

        .feature-list li::before {
          content: '';
          position: absolute;
          left: 2px;
          top: 50%;
          transform: translateY(-50%) rotate(45deg);
          width: 5px;
          height: 9px;
          border-right: 1.5px solid var(--color-success);
          border-bottom: 1.5px solid var(--color-success);
        }

        .resources {
          max-width: 800px;
          margin: 0 auto;
        }

        .resources h3 {
          text-align: center;
          margin-bottom: var(--space-lg);
          color: var(--color-text-secondary);
        }

        .resources-grid {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(200px, 1fr));
          gap: var(--space-md);
        }

        .resource-link {
          display: flex;
          align-items: center;
          gap: var(--space-md);
          padding: var(--space-md);
          background: var(--color-bg);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: var(--radius-md);
          color: var(--color-text);
          transition: all var(--transition-fast);
        }

        .resource-link:hover {
          border-color: var(--color-accent);
          color: var(--color-text);
        }

        .resource-icon {
          color: var(--color-accent);
        }

        .resource-text {
          display: flex;
          flex-direction: column;
        }

        .resource-text strong {
          font-size: 0.875rem;
        }

        .resource-text span:last-child {
          font-size: 0.75rem;
          color: var(--color-text-muted);
        }
      `}</style>
    </section>
  );
});
