// @a0n/gnosis/forkjoin-app/client/components/Sections/LatencyRace.tsx
/**
 * LatencyRace - Animated Latency Comparison
 *
 * Dots bouncing back and forth at different speeds.
 * We're bouncing madly. They're crawling.
 *
 * Displayed in nanoseconds for precision.
 * Animation durations explicitly set for visual clarity.
 */

import { memo } from '@a0n/raect';

interface RacerProps {
  name: string;
  latencyNs: number; // Latency in nanoseconds
  durationS: number; // Animation duration in seconds (explicit control)
  color: string;
  description: string;
}

// Format nanoseconds for display (e.g., 100K, 7M, 150M)
function formatNs(ns: number): string {
  if (ns >= 1_000_000_000) {
    return `${(ns / 1_000_000_000).toFixed(1)}B`;
  } else if (ns >= 1_000_000) {
    return `${(ns / 1_000_000).toFixed(0)}M`;
  } else if (ns >= 1_000) {
    return `${(ns / 1_000).toFixed(0)}K`;
  }
  return `${ns}`;
}

const Racer = memo(function Racer({
  name,
  latencyNs,
  durationS,
  color,
  description,
}: RacerProps) {
  return (
    <div className="racer">
      <div className="racer-label">
        <span className="racer-name">{name}</span>
        <span className="racer-latency" style={{ color }}>
          {formatNs(latencyNs)} ns
        </span>
      </div>
      <div className="racer-track">
        <div
          className="racer-dot"
          style={{
            backgroundColor: color,
            animationDuration: `${durationS}s`,
            boxShadow: `0 0 10px ${color}, 0 0 20px ${color}40`,
          }}
        />
        <div
          className="racer-trail"
          style={{
            background: `linear-gradient(90deg, transparent, ${color}20)`,
          }}
        />
      </div>
      <div className="racer-desc">{description}</div>
    </div>
  );
});

export const LatencyRace = memo(function LatencyRace() {
  // Latency in nanoseconds, with explicit animation durations for visual distinction
  // Slower durationS = slower dots = higher latency
  const racers: RacerProps[] = [
    {
      name: 'Aeon Foundation',
      latencyNs: 100_000, // 0.1ms = 100K ns
      durationS: 0.6, // Fastest - visibly quick
      color: '#10b981',
      description: 'KV cache + WASM',
    },
    {
      name: 'Edge Render',
      latencyNs: 7_000_000, // 7ms = 7M ns
      durationS: 1.4, // Noticeably slower than Aeon
      color: '#06b6d4',
      description: 'D1 + Workers',
    },
    {
      name: 'Session Cache',
      latencyNs: 50_000_000, // 50ms = 50M ns
      durationS: 3.0, // Getting sluggish
      color: '#5599cc',
      description: 'Full page render',
    },
    {
      name: 'Major Framework',
      latencyNs: 150_000_000, // 150ms = 150M ns
      durationS: 5.5, // Slow
      color: '#ef4444',
      description: 'Origin server',
    },
    {
      name: 'Traditional SSR',
      latencyNs: 300_000_000, // 300ms = 300M ns
      durationS: 9.0, // Painfully slow
      color: '#f59e0b',
      description: 'Cold start + DB',
    },
  ];

  return (
    <div className="latency-race">
      <div className="race-header">
        <h3>Request → Response</h3>
        <p>Measured in nanoseconds. Watch the dots.</p>
      </div>

      <div className="race-track-container">
        {racers.map((racer) => (
          <Racer key={racer.name} {...racer} />
        ))}
      </div>

      <div className="race-legend">
        <span className="legend-fast">← Faster</span>
        <span className="legend-slow">Slower →</span>
      </div>

      <style>{`
        .latency-race {
          max-width: 700px;
          margin: 0 auto var(--space-3xl);
          padding: var(--space-xl);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 16px;
        }

        .race-header {
          text-align: center;
          margin-bottom: var(--space-xl);
        }

        .race-header h3 {
          margin-bottom: var(--space-xs);
          color: var(--color-text);
        }

        .race-header p {
          font-size: 0.875rem;
          color: var(--color-text-muted);
          margin: 0;
        }

        .race-track-container {
          display: flex;
          flex-direction: column;
          gap: var(--space-md);
        }

        .racer {
          display: grid;
          grid-template-columns: 140px 1fr 100px;
          align-items: center;
          gap: var(--space-md);
        }

        .racer-label {
          display: flex;
          flex-direction: column;
        }

        .racer-name {
          font-size: 0.85rem;
          font-weight: 500;
          color: var(--color-text);
        }

        .racer-latency {
          font-family: var(--font-mono);
          font-size: 0.75rem;
          font-weight: 600;
        }

        .racer-track {
          position: relative;
          height: 24px;
          background: var(--color-bg-tertiary);
          border-radius: 12px;
          overflow: hidden;
        }

        .racer-trail {
          position: absolute;
          inset: 0;
          opacity: 0.5;
        }

        .racer-dot {
          position: absolute;
          top: 50%;
          left: 4px;
          width: 16px;
          height: 16px;
          border-radius: 50%;
          transform: translateY(-50%);
          animation: race-bounce linear infinite;
        }

        @keyframes race-bounce {
          0%, 100% {
            left: 4px;
          }
          50% {
            left: calc(100% - 20px);
          }
        }

        .racer-desc {
          font-size: 0.7rem;
          color: var(--color-text-muted);
          text-align: right;
        }

        .race-legend {
          display: flex;
          justify-content: space-between;
          margin-top: var(--space-lg);
          padding-top: var(--space-md);
          border-top: 1px solid var(--color-bg-tertiary);
        }

        .legend-fast {
          font-size: 0.7rem;
          color: #10b981;
          text-transform: uppercase;
          letter-spacing: 0.1em;
        }

        .legend-slow {
          font-size: 0.7rem;
          color: #ef4444;
          text-transform: uppercase;
          letter-spacing: 0.1em;
        }

        @media (max-width: 600px) {
          .racer {
            grid-template-columns: 1fr;
            gap: var(--space-xs);
          }

          .racer-label {
            flex-direction: row;
            justify-content: space-between;
          }

          .racer-desc {
            text-align: left;
          }
        }
      `}</style>
    </div>
  );
});
