// @a0n/gnosis/forkjoin-app/client/components/Sections/NeuralSection.tsx
/**
 * NeuralSection - Distributed Learning & Collective Wisdom
 *
 * Privacy-preserving AI that learns from the collective:
 * - WebNN-first inference with WebGPU fallback
 * - Federated learning (gradients only leave the edge)
 * - Differential privacy (mathematical guarantees)
 * - Collective intelligence (patterns, not data)
 */

import {
  Shield,
  Brain,
  Users,
  Cpu,
  Zap,
  Lock,
  Eye,
  Sparkles,
  Layers,
  Network,
  Activity,
  BarChart3,
  Fingerprint,
  Globe,
  Gauge,
  Share2,
} from '@a0n/aeon-icons';

const features = [
  // Row 1: Inference
  {
    icon: Cpu,
    title: 'WebNN First',
    metric: 'NPU',
    description:
      'Apple Neural Engine, Qualcomm Hexagon. Hardware AI on any device.',
    category: 'inference',
  },
  {
    icon: Zap,
    title: 'WebGPU Fallback',
    metric: '1000x',
    description: 'High-performance GPU parallelization when NPU unavailable.',
    category: 'inference',
  },
  {
    icon: Gauge,
    title: 'WASM Backend',
    metric: 'Universal',
    description: 'SIMD-optimized WASM for maximum browser compatibility.',
    category: 'inference',
  },
  {
    icon: Activity,
    title: 'Streaming Output',
    metric: 'Real-time',
    description: 'Token-by-token streaming. See results as they generate.',
    category: 'inference',
  },

  // Row 2: Federated Learning
  {
    icon: Network,
    title: 'Federated Learning',
    metric: 'Edge',
    description: 'Train on user devices. Only gradients leave, never raw data.',
    category: 'federated',
  },
  {
    icon: Share2,
    title: 'Gradient Sharing',
    metric: 'Compress',
    description: 'Compressed gradient updates. 90% bandwidth reduction.',
    category: 'federated',
  },
  {
    icon: Layers,
    title: 'Model Aggregation',
    metric: 'FedAvg',
    description:
      'Federated averaging for global model updates. Secure aggregation.',
    category: 'federated',
  },
  {
    icon: Globe,
    title: 'Distributed Training',
    metric: 'Global',
    description:
      'Models improve from worldwide usage. Collective gets smarter.',
    category: 'federated',
  },

  // Row 3: Differential Privacy
  {
    icon: Shield,
    title: 'Differential Privacy',
    metric: 'Math',
    description:
      'Mathematical guarantees. Individual contributions stay hidden.',
    category: 'privacy',
  },
  {
    icon: Fingerprint,
    title: 'K-Anonymity',
    metric: '10+',
    description:
      'Minimum cohort size before any data is shared. Strength in numbers.',
    category: 'privacy',
  },
  {
    icon: Eye,
    title: 'Noise Injection',
    metric: '+/-0.1',
    description:
      'Calibrated noise added to all outputs. Plausible deniability.',
    category: 'privacy',
  },
  {
    icon: Lock,
    title: 'Secure Enclaves',
    metric: 'TEE',
    description: 'Trusted execution when available. Hardware-level isolation.',
    category: 'privacy',
  },

  // Row 4: Collective Intelligence
  {
    icon: Brain,
    title: 'Pattern Learning',
    metric: 'Emerge',
    description: 'Learn from patterns that emerge. Never individual behavior.',
    category: 'collective',
  },
  {
    icon: Users,
    title: 'Collective Wisdom',
    metric: 'Aggregate',
    description: 'Wisdom of the crowd without surveillance. Privacy by design.',
    category: 'collective',
  },
  {
    icon: BarChart3,
    title: 'Behavioral Signals',
    metric: 'Cursor',
    description:
      'Mouse movements, dwell time, scroll patterns. All anonymized.',
    category: 'collective',
  },
  {
    icon: Sparkles,
    title: 'Emotional State',
    metric: 'Infer',
    description:
      'Collective emotion from aggregate behavior. No individual exposed.',
    category: 'collective',
  },
];

const categoryColors: Record<string, string> = {
  inference: '#5599cc',
  federated: '#10b981',
  privacy: '#f59e0b',
  collective: '#06b6d4',
};

/**
 * Renders the Neural Section view.
 */
export function NeuralSection() {
  return (
    <section
      id="neural"
      className="section neural-section"
      aria-labelledby="neural-title"
    >
      <div className="container">
        <div className="section-header">
          <span className="layer-label">Intelligence Layer</span>
          <h2 id="neural-title" className="product-section-title">
            Neural
          </h2>
          <p className="section-description">
            WebNN-first inference with WebGPU fallback. Distributed learning
            that respects privacy.
          </p>
        </div>

        <div className="neural-grid">
          {features.map((feature) => (
            <div
              key={feature.title}
              className="neural-tile"
              style={
                {
                  '--tile-color': categoryColors[feature.category],
                } as React.CSSProperties
              }
            >
              <feature.icon size={24} className="tile-icon" />
              <div className="tile-content">
                <div className="tile-header">
                  <h3>{feature.title}</h3>
                  <span className="tile-metric">{feature.metric}</span>
                </div>
                <p>{feature.description}</p>
              </div>
            </div>
          ))}
        </div>

        <div className="neural-demo">
          <h3>This Page Uses Neural</h3>
          <div className="demo-metrics">
            <div className="metric">
              <span className="metric-value">10+</span>
              <span className="metric-label">
                Visitors needed for emotion data
              </span>
            </div>
            <div className="metric">
              <span className="metric-value">+/-0.1</span>
              <span className="metric-label">Noise added for privacy</span>
            </div>
            <div className="metric">
              <span className="metric-value">0</span>
              <span className="metric-label">
                Individual data points exposed
              </span>
            </div>
          </div>
          <p className="demo-explanation">
            The collective emotion you see in the visualization comes from
            aggregating everyone's cursor movements, interaction patterns, and
            time spent on sections. But no individual's behavior is ever exposed
            - only the collective emerges.
          </p>
        </div>

        <div className="neural-principles">
          <h3>Core Principles</h3>
          <div className="principles-grid">
            <div className="principle">
              <span className="principle-num">01</span>
              <p>WebNN first - runs on any device with a neural accelerator.</p>
            </div>
            <div className="principle">
              <span className="principle-num">02</span>
              <p>Data stays on the edge. Models come to you.</p>
            </div>
            <div className="principle">
              <span className="principle-num">03</span>
              <p>Privacy is not a feature. It's the architecture.</p>
            </div>
            <div className="principle">
              <span className="principle-num">04</span>
              <p>Collective wisdom without collective surveillance.</p>
            </div>
          </div>
        </div>

        <div className="cta-row">
          <a href="#/neural" className="button button-primary">
            Learn More
          </a>
        </div>
      </div>

      <style>{`
        .neural-section {
          background: linear-gradient(180deg, var(--color-bg-secondary) 0%, #0f0a1a 100%);
        }

        .section-header {
          text-align: center;
          margin-bottom: var(--space-3xl);
        }

        .section-description {
          font-size: 1.125rem;
          color: var(--color-text-secondary);
          max-width: 600px;
          margin: var(--space-md) auto 0;
        }

        .neural-grid {
          display: grid;
          grid-template-columns: repeat(4, 1fr);
          gap: var(--space-md);
          margin-bottom: var(--space-3xl);
        }

        .neural-tile {
          display: flex;
          gap: var(--space-md);
          padding: var(--space-md);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 10px;
          transition: all 0.2s ease;
        }

        .neural-tile:hover {
          border-color: var(--tile-color);
          transform: translateY(-2px);
        }

        .tile-icon {
          color: var(--tile-color);
          flex-shrink: 0;
          margin-top: 2px;
        }

        .tile-content {
          flex: 1;
          min-width: 0;
        }

        .tile-header {
          display: flex;
          align-items: center;
          justify-content: space-between;
          gap: var(--space-sm);
          margin-bottom: var(--space-xs);
        }

        .tile-header h3 {
          font-size: 0.9rem;
          font-weight: 600;
          color: var(--color-text);
          margin: 0;
          white-space: nowrap;
          overflow: hidden;
          text-overflow: ellipsis;
        }

        .tile-metric {
          font-family: var(--font-mono);
          font-size: 0.7rem;
          font-weight: 600;
          color: var(--tile-color);
          padding: 2px 6px;
          background: rgba(255, 255, 255, 0.05);
          border-radius: 4px;
          flex-shrink: 0;
        }

        .neural-tile p {
          font-size: 0.8rem;
          color: var(--color-text-muted);
          margin: 0;
          line-height: 1.4;
        }

        .neural-demo {
          max-width: 700px;
          margin: 0 auto var(--space-3xl);
          text-align: center;
          padding: var(--space-xl);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 16px;
        }

        .neural-demo h3 {
          margin-bottom: var(--space-lg);
          color: var(--color-accent);
        }

        .demo-metrics {
          display: flex;
          justify-content: center;
          gap: var(--space-xl);
          margin-bottom: var(--space-lg);
          flex-wrap: wrap;
        }

        .metric {
          display: flex;
          flex-direction: column;
          align-items: center;
        }

        .metric-value {
          font-size: 2rem;
          font-weight: 700;
          color: #5599cc;
        }

        .metric-label {
          font-size: 0.75rem;
          color: var(--color-text-muted);
          text-transform: uppercase;
          letter-spacing: 0.05em;
        }

        .demo-explanation {
          font-size: 0.9rem;
          color: var(--color-text-secondary);
          margin: 0 auto;
          max-width: 600px;
          text-align: center;
          line-height: 1.6;
        }

        .neural-principles {
          max-width: 700px;
          margin: 0 auto var(--space-xl);
        }

        .neural-principles h3 {
          text-align: center;
          margin-bottom: var(--space-lg);
          color: var(--color-text-muted);
          text-transform: uppercase;
          letter-spacing: 0.1em;
          font-size: 0.85rem;
        }

        .principles-grid {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(250px, 1fr));
          gap: var(--space-md);
        }

        .principle {
          display: flex;
          align-items: flex-start;
          gap: var(--space-sm);
          padding: var(--space-md);
          background: rgba(51, 102, 153, 0.05);
          border-left: 3px solid #5599cc;
          border-radius: 0 8px 8px 0;
        }

        .principle-num {
          font-family: var(--font-mono);
          font-size: 0.8rem;
          color: #5599cc;
          opacity: 0.7;
        }

        .principle p {
          margin: 0;
          font-size: 0.9rem;
          color: var(--color-text-secondary);
        }

        .cta-row {
          display: flex;
          justify-content: center;
        }

        @media (max-width: 1200px) {
          .neural-grid {
            grid-template-columns: repeat(2, 1fr);
          }
        }

        @media (max-width: 768px) {
          .demo-metrics {
            flex-direction: column;
            gap: var(--space-md);
          }
        }

        @media (max-width: 600px) {
          .neural-grid {
            grid-template-columns: 1fr;
          }

          .neural-tile {
            flex-direction: column;
            text-align: center;
          }

          .tile-header {
            justify-content: center;
          }

          .tile-icon {
            margin: 0 auto;
          }
        }
      `}</style>
    </section>
  );
}
