// @a0n/gnosis/forkjoin-app/client/components/Sections/PlatformsSection.tsx
/**
 * PlatformsSection - Platform targets, driven by build-time manifest.
 */

import { ArrowRight } from '@a0n/aeon-icons';
import manifest from '../../generated/ecosystem-manifest.json';

/**
 * Renders the Platforms Section view.
 */
export function PlatformsSection() {
  const platforms = manifest.platforms;

  return (
    <section
      id="platforms"
      className="section platforms-section"
      aria-labelledby="platforms-title"
    >
      <div className="container">
        <div className="section-header">
          <h2 id="platforms-title">Every Platform</h2>
          <p className="section-description">
            One stack. {platforms.length} surfaces. Local-first everywhere.
          </p>
        </div>

        <div className="platforms-grid">
          {platforms.map((platform) => (
            <a
              key={platform.repo}
              href={`https://github.com/forkjoin-ai/${platform.repo}`}
              target="_blank"
              rel="noopener noreferrer"
              className="platform-card"
            >
              <div className="platform-card-header">
                <span className="platform-card-name">{platform.name}</span>
                <span className="platform-card-runtime">
                  {platform.runtime}
                </span>
                <ArrowRight size={14} className="platform-card-arrow" />
              </div>
            </a>
          ))}
        </div>
      </div>

      <style>{`
        .platforms-section {
          background: linear-gradient(180deg, #0a0a12 0%, #12121a 50%, #0a0a12 100%);
        }

        .platforms-grid {
          display: grid;
          grid-template-columns: repeat(auto-fill, minmax(260px, 1fr));
          gap: var(--space-md);
        }

        .platform-card {
          display: flex;
          flex-direction: column;
          gap: 0.5rem;
          padding: var(--space-md) var(--space-lg);
          border-radius: 12px;
          background: rgba(255, 255, 255, 0.03);
          border: 1px solid rgba(255, 255, 255, 0.08);
          text-decoration: none;
          color: inherit;
          transition: transform 0.2s ease, border-color 0.2s ease;
        }

        .platform-card:hover {
          transform: translateY(-2px);
          border-color: var(--color-accent);
        }

        .platform-card-header {
          display: flex;
          align-items: center;
          gap: var(--space-sm);
        }

        .platform-card-name {
          font-weight: 600;
          font-size: 1.1rem;
          color: var(--color-accent);
          flex: 1;
        }

        .platform-card-runtime {
          font-size: 0.7rem;
          text-transform: uppercase;
          letter-spacing: 0.08em;
          padding: 2px 8px;
          border-radius: 4px;
          background: rgba(59, 130, 246, 0.15);
          color: var(--color-accent);
        }

        .platform-card-arrow {
          color: var(--color-accent);
          opacity: 0;
          transition: opacity 0.2s ease;
        }

        .platform-card:hover .platform-card-arrow {
          opacity: 1;
        }
      `}</style>
    </section>
  );
}
