// @a0n/.ignored_gnosis/src/crdt/qdoc.ts
'use aeon';

/**
 * QDoc — Quantum Document: the topology-native replacement for Y.Doc
 *
 * CRDT is the only state model. No memory. No GC. Append-only.
 * The topology formalizes the state. FORK creates superposition.
 * OBSERVE collapses. History is permanent.
 *
 * QDoc wraps the GG topology model checker and provides:
 *   - QMap, QArray, QText, QCounter — typed CRDT accessors
 *   - Topology delta encoding (replaces Y.encodeStateAsUpdate)
 *   - Topology delta application (replaces Y.applyUpdate)
 *   - Observable changes (replaces Y.Doc.on('update'))
 *   - SLIVER-based presence (replaces Awareness)
 *
 * The internal representation is a GG topology. Every mutation
 * appends a FORK branch. Every read triggers OBSERVE (collapse).
 * The topology is append-only — no tombstones, no GC.
 */

type GgCollapseStrategy = string;

// ── Types ───────────────────────────────────────────────────────────────────

export interface QDocOptions {
  /** Unique document ID */
  guid?: string;
  /** Default collapse strategy for maps */
  mapStrategy?: GgCollapseStrategy;
  /** Default collapse strategy for sequences */
  sequenceStrategy?: GgCollapseStrategy;
}

export interface QDocDelta {
  /** Topology operations since last sync */
  nodes: readonly QDocDeltaNode[];
  edges: readonly QDocDeltaEdge[];
  /** Logical clock for causal ordering */
  clock: number;
  /** Source replica ID */
  replicaId: string;
}

export interface QDocCanonicalSnapshot {
  format: 'qdoc.canonical-snapshot.v1';
  guid: string;
  maps: Record<string, Record<string, unknown>>;
  arrays: Record<string, unknown[]>;
  texts: Record<string, string>;
  counters: Record<string, number>;
}

export interface QDocDeltaNode {
  id: string;
  labels: readonly string[];
  properties: Readonly<Record<string, string>>;
}

export interface QDocDeltaEdge {
  sourceIds: readonly string[];
  targetIds: readonly string[];
  type: string;
  properties: Readonly<Record<string, string>>;
}

export type QDocUpdateHandler = (delta: Uint8Array, origin: string) => void;
export type QDocObserveHandler = (event: QDocEvent) => void;

export interface QDocEvent {
  type: 'insert' | 'delete' | 'set' | 'observe';
  path: string;
  key?: string;
  value?: unknown;
  previousValue?: unknown;
  origin: string;
}

export interface QMapKeyChange<T = unknown> {
  action: 'add' | 'update' | 'delete';
  oldValue?: T;
}

export interface QMapEvent<T = unknown> {
  keysChanged: Set<string>;
  changes: {
    keys: Map<string, QMapKeyChange<T>>;
  };
}

interface SerializedCrdtEnvelope {
  __qdocType: 'string' | 'json' | 'undefined';
  value?: unknown;
}

function isSerializedCrdtEnvelope(
  value: unknown
): value is SerializedCrdtEnvelope {
  return (
    typeof value === 'object' &&
    value !== null &&
    '__qdocType' in value &&
    (((value as SerializedCrdtEnvelope).__qdocType === 'string' &&
      'value' in value) ||
      ((value as SerializedCrdtEnvelope).__qdocType === 'json' &&
        'value' in value) ||
      (value as SerializedCrdtEnvelope).__qdocType === 'undefined')
  );
}

function serializeCrdtValue(value: unknown): string {
  if (typeof value === 'string') {
    return JSON.stringify({
      __qdocType: 'string',
      value,
    } satisfies SerializedCrdtEnvelope);
  }

  if (typeof value === 'undefined') {
    return JSON.stringify({
      __qdocType: 'undefined',
    } satisfies SerializedCrdtEnvelope);
  }

  return JSON.stringify({
    __qdocType: 'json',
    value,
  } satisfies SerializedCrdtEnvelope);
}

function deserializeCrdtValue(serialized: string | undefined): unknown {
  if (typeof serialized !== 'string') {
    return undefined;
  }

  try {
    const parsed = JSON.parse(serialized) as unknown;
    if (isSerializedCrdtEnvelope(parsed)) {
      return parsed.__qdocType === 'undefined' ? undefined : parsed.value;
    }
    if (
      parsed === null ||
      Array.isArray(parsed) ||
      (typeof parsed === 'object' && parsed !== null)
    ) {
      return parsed;
    }
  } catch {
    return serialized;
  }

  return serialized;
}

// ── Topology Events ─────────────────────────────────────────────────────────

export type QDocTopologyEventKind = 'fork' | 'observe' | 'fold';

export interface QDocTopologyEvent {
  kind: QDocTopologyEventKind;
  path: string;
  sourceIds: string[];
  targetIds: string[];
  timestamp: number;
}

// ── QDoc ────────────────────────────────────────────────────────────────────

export class QDoc {
  readonly guid: string;
  private readonly _replicaId: string;
  private _clock = 0;
  private _beta1 = 0;

  // Append-only topology state — this formalizes the document
  private _nodes: Map<string, QDocDeltaNode> = new Map();
  private _edges: QDocDeltaEdge[] = [];

  // Pending deltas not yet synced
  private _pendingNodes: QDocDeltaNode[] = [];
  private _pendingEdges: QDocDeltaEdge[] = [];

  // Observable
  private _updateHandlers: Set<QDocUpdateHandler> = new Set();
  private _observeHandlers: Map<string, Set<QDocObserveHandler>> = new Map();

  // Typed accessors (lazy, cached)
  private _maps: Map<string, QMap<unknown>> = new Map();
  private _arrays: Map<string, QArray> = new Map();
  private _texts: Map<string, QText> = new Map();
  private _xmlFragments: Map<string, QXmlFragment> = new Map();
  private _counters: Map<string, QCounter> = new Map();

  // Presence (SLIVER — never collapses)
  private _presence: Map<string, Record<string, unknown>> = new Map();
  private _presenceHandlers: Set<
    (states: Map<string, Record<string, unknown>>) => void
  > = new Set();

  private readonly _mapStrategy: GgCollapseStrategy;
  private readonly _sequenceStrategy: GgCollapseStrategy;

  // Topology event handlers
  private _topologyHandlers: Set<(event: QDocTopologyEvent) => void> =
    new Set();

  constructor(options: QDocOptions = {}) {
    this.guid = options.guid ?? crypto.randomUUID();
    this._replicaId = `replica-${crypto.randomUUID().slice(0, 8)}`;
    this._mapStrategy = options.mapStrategy ?? 'lww';
    this._sequenceStrategy = options.sequenceStrategy ?? 'ot-transform';

    // Root node — every topology has one
    this._appendNode({
      id: 'root',
      labels: ['QDoc'],
      properties: { guid: this.guid },
    });
  }

  // ── Typed Accessors (replaces Y.Doc.getMap, getArray, getText) ──────────

  getMap<T = unknown>(name: string): QMap<T> {
    let map = this._maps.get(name);
    if (!map) {
      const hasTopology = this._nodes.has(`map_${name}`);
      map = new QMap(this, name, this._mapStrategy, !hasTopology);
      if (hasTopology) {
        this._replayMapState(name, map);
      }
      this._maps.set(name, map);
    }
    return map as QMap<T>;
  }

  getArray<T = unknown>(name: string): QArray<T> {
    let arr = this._arrays.get(name);
    if (!arr) {
      const hasTopology = this._nodes.has(`arr_${name}`);
      arr = new QArray(this, name, this._sequenceStrategy, !hasTopology);
      if (hasTopology) {
        this._replayArrayState(name, arr);
      }
      this._arrays.set(name, arr);
    }
    return arr as QArray<T>;
  }

  getText(name: string): QText {
    let text = this._texts.get(name);
    if (!text) {
      const hasTopology = this._nodes.has(`text_${name}`);
      text = new QText(this, name, !hasTopology);
      if (hasTopology) {
        this._replayTextState(name, text);
      }
      this._texts.set(name, text);
    }
    return text;
  }

  getXmlFragment(name: string): QXmlFragment {
    let fragment = this._xmlFragments.get(name);
    if (!fragment) {
      fragment = new QXmlFragment(this, name);
      this._xmlFragments.set(name, fragment);
    }
    return fragment;
  }

  getCounter(name: string): QCounter {
    let counter = this._counters.get(name);
    if (!counter) {
      const hasTopology = this._nodes.has(`ctr_${name}`);
      counter = new QCounter(this, name, !hasTopology);
      if (hasTopology) {
        this._replayCounterState(name, counter);
      }
      this._counters.set(name, counter);
    }
    return counter;
  }

  /**
   * Execute a function in a transaction scope.
   * In QDoc, every operation is atomic (append-only topology), so this
   * is a compatibility shim for Y.Doc.transact() migration.
   * The function runs immediately and notifications fire after.
   */
  transact(fn: () => void, _origin = 'local'): void {
    fn();
  }

  /**
   * Destroy the document (compatibility with Y.Doc.destroy()).
   * Clears handlers. The topology itself is immutable/append-only.
   */
  destroy(): void {
    this._updateHandlers.clear();
    this._observeHandlers.clear();
    this._presenceHandlers.clear();
    this._topologyHandlers.clear();
  }

  // ── Topology Mutation (internal — called by typed accessors) ────────────

  /** @internal Append a node to the topology */
  _appendNode(node: QDocDeltaNode): void {
    this._nodes.set(node.id, node);
    this._pendingNodes.push(node);
  }

  /** @internal Append an edge to the topology */
  _appendEdge(edge: QDocDeltaEdge): void {
    this._edges.push(edge);
    this._pendingEdges.push(edge);

    // Update beta1
    if (edge.type === 'FORK') {
      this._beta1 += edge.targetIds.length - 1;
    } else if (
      edge.type === 'FOLD' ||
      edge.type === 'COLLAPSE' ||
      edge.type === 'OBSERVE'
    ) {
      this._beta1 = 0;
    } else if (edge.type === 'RACE') {
      this._beta1 = Math.max(0, this._beta1 - (edge.sourceIds.length - 1));
    } else if (edge.type === 'VENT' || edge.type === 'TUNNEL') {
      this._beta1 = Math.max(0, this._beta1 - 1);
    }

    this._clock++;

    // Emit topology event for observable edge types
    const path = edge.properties.path ?? '';
    if (edge.type === 'FORK') {
      this.emitTopologyEvent({
        kind: 'fork',
        path,
        sourceIds: [...edge.sourceIds],
        targetIds: [...edge.targetIds],
        timestamp: Date.now(),
      });
    } else if (edge.type === 'OBSERVE' || edge.type === 'RACE') {
      this.emitTopologyEvent({
        kind: 'observe',
        path,
        sourceIds: [...edge.sourceIds],
        targetIds: [...edge.targetIds],
        timestamp: Date.now(),
      });
    } else if (edge.type === 'FOLD' || edge.type === 'COLLAPSE') {
      this.emitTopologyEvent({
        kind: 'fold',
        path,
        sourceIds: [...edge.sourceIds],
        targetIds: [...edge.targetIds],
        timestamp: Date.now(),
      });
    }
  }

  /** @internal Get the current beta1 (superposition count) */
  get beta1(): number {
    return this._beta1;
  }

  /** @internal Get the logical clock */
  get clock(): number {
    return this._clock;
  }

  /** @internal Get the replica ID */
  get replicaId(): string {
    return this._replicaId;
  }

  /** @internal Emit an observe event */
  _emitObserve(path: string, event: QDocEvent): void {
    const handlers = this._observeHandlers.get(path);
    if (handlers) {
      for (const handler of handlers) {
        try {
          handler(event);
        } catch {
          /* handlers must not break the doc */
        }
      }
    }
  }

  /** @internal Notify update handlers */
  _notifyUpdate(origin: string, delta?: Uint8Array): void {
    const emittedDelta =
      delta ??
      (this._pendingNodes.length === 0 && this._pendingEdges.length === 0
        ? null
        : this.encodePendingDelta());
    if (!emittedDelta) {
      return;
    }
    for (const handler of this._updateHandlers) {
      try {
        handler(emittedDelta, origin);
      } catch {
        /* handlers must not break the doc */
      }
    }
  }

  // ── Topology Events ──────────────────────────────────────────────────

  /** Subscribe to topology events (fork/observe/fold). Returns unsubscribe function. */
  public onTopologyEvent(
    handler: (event: QDocTopologyEvent) => void
  ): () => void {
    this._topologyHandlers.add(handler);
    return () => this._topologyHandlers.delete(handler);
  }

  /** @internal Emit a topology event to all handlers */
  private emitTopologyEvent(event: QDocTopologyEvent): void {
    for (const handler of this._topologyHandlers) {
      handler(event);
    }
  }

  // ── Sync Protocol (replaces Y.encodeStateAsUpdate / Y.applyUpdate) ─────

  /**
   * Encode the full topology state as a Uint8Array.
   * Replaces Y.encodeStateAsUpdate(doc).
   */
  encodeStateAsUpdate(_stateVector?: Uint8Array): Uint8Array {
    const state: QDocDelta = {
      nodes: [...this._nodes.values()],
      edges: this._edges,
      clock: this._clock,
      replicaId: this._replicaId,
    };
    return new TextEncoder().encode(JSON.stringify(state));
  }

  /**
   * Encode a collapsed snapshot of the current visible state.
   * This strips append-only write history so local persistence does not
   * replay stale churn on every restore.
   */
  encodeCanonicalStateAsUpdate(): Uint8Array {
    const canonical = new QDoc({
      guid: this.guid,
      mapStrategy: this._mapStrategy,
      sequenceStrategy: this._sequenceStrategy,
    });

    for (const name of this._topLevelCollectionNames('map_')) {
      const source = this.getMap(name);
      const target = canonical.getMap(name);
      for (const [key, value] of source.entries()) {
        target.set(key, value);
      }
    }

    for (const name of this._topLevelCollectionNames('arr_')) {
      const values = this.getArray(name).toArray();
      const target = canonical.getArray(name);
      if (values.length > 0) {
        target.push(values);
      }
    }

    for (const name of this._topLevelCollectionNames('text_')) {
      const value = this.getText(name).toString();
      const target = canonical.getText(name);
      if (value.length > 0) {
        target.insert(0, value);
      }
    }

    for (const name of this._topLevelCollectionNames('ctr_')) {
      const value = this.getCounter(name).value;
      const target = canonical.getCounter(name);
      if (value !== 0) {
        target.increment(value);
      }
    }

    return canonical.encodeStateAsUpdate();
  }

  serializeCanonicalSnapshot(): QDocCanonicalSnapshot {
    const maps: Record<string, Record<string, unknown>> = {};
    for (const name of this._topLevelCollectionNames('map_')) {
      maps[name] = this.getMap(name).toJSON();
    }

    const arrays: Record<string, unknown[]> = {};
    for (const name of this._topLevelCollectionNames('arr_')) {
      arrays[name] = this.getArray(name).toJSON();
    }

    const texts: Record<string, string> = {};
    for (const name of this._topLevelCollectionNames('text_')) {
      texts[name] = this.getText(name).toString();
    }

    const counters: Record<string, number> = {};
    for (const name of this._topLevelCollectionNames('ctr_')) {
      counters[name] = this.getCounter(name).value;
    }

    return {
      format: 'qdoc.canonical-snapshot.v1',
      guid: this.guid,
      maps,
      arrays,
      texts,
      counters,
    };
  }

  applyCanonicalSnapshot(snapshot: QDocCanonicalSnapshot): void {
    for (const [name, entries] of Object.entries(snapshot.maps)) {
      const map = this.getMap(name);
      for (const [key, value] of Object.entries(entries)) {
        map.set(key, value);
      }
    }

    for (const [name, values] of Object.entries(snapshot.arrays)) {
      if (values.length === 0) {
        this.getArray(name);
        continue;
      }
      this.getArray(name).push(values);
    }

    for (const [name, value] of Object.entries(snapshot.texts)) {
      const text = this.getText(name);
      if (value.length > 0) {
        text.insert(0, value);
      }
    }

    for (const [name, value] of Object.entries(snapshot.counters)) {
      if (value === 0) {
        this.getCounter(name);
        continue;
      }
      this.getCounter(name).increment(value);
    }
  }

  private _topLevelCollectionNames(prefix: string): string[] {
    const names = new Set<string>();
    for (const node of this._nodes.values()) {
      if (node.id.startsWith(prefix)) {
        names.add(node.id.slice(prefix.length));
      }
    }
    return [...names].sort();
  }

  encodeStateVector(): Uint8Array {
    return new Uint8Array(0);
  }

  /**
   * Encode only pending (unsynced) operations.
   * More efficient than full state for incremental sync.
   */
  encodePendingDelta(): Uint8Array {
    const delta: QDocDelta = {
      nodes: this._pendingNodes,
      edges: this._pendingEdges,
      clock: this._clock,
      replicaId: this._replicaId,
    };
    const encoded = new TextEncoder().encode(JSON.stringify(delta));
    // Clear pending after encoding
    this._pendingNodes = [];
    this._pendingEdges = [];
    return encoded;
  }

  /**
   * Apply a remote topology delta.
   * Replaces Y.applyUpdate(doc, update).
   * Append-only — remote operations are FORK branches that get merged via OBSERVE.
   */
  applyUpdate(update: Uint8Array, origin = 'remote'): void {
    let delta: QDocDelta;
    try {
      delta = JSON.parse(
        new TextDecoder('utf-8', { fatal: true, ignoreBOM: false }).decode(
          update
        )
      );
    } catch {
      return; // Invalid update — ignore
    }

    // Append remote nodes (append-only, never overwrite)
    for (const node of delta.nodes) {
      if (!this._nodes.has(node.id)) {
        this._nodes.set(node.id, node);
      }
    }

    // Append remote edges (the topology only grows)
    for (const edge of delta.edges) {
      this._edges.push(edge);

      // Update beta1 for remote edges
      if (edge.type === 'FORK') {
        this._beta1 += edge.targetIds.length - 1;
      } else if (
        edge.type === 'FOLD' ||
        edge.type === 'COLLAPSE' ||
        edge.type === 'OBSERVE'
      ) {
        this._beta1 = 0;
      }
    }

    // Advance clock to max of local and remote
    if (delta.clock > this._clock) {
      this._clock = delta.clock;
    }

    for (const edge of delta.edges) {
      const event = this._applyRemoteEdge(edge, origin);
      if (event) {
        this._emitObserve(event.path, event);
      }
    }

    this._notifyUpdate(origin, update);
  }

  private _applyRemoteEdge(
    edge: QDocDeltaEdge,
    origin: string
  ): QDocEvent | null {
    const path = edge.properties.path;
    const sourceId = edge.sourceIds[0] ?? '';
    if (!path) {
      return null;
    }

    if (
      edge.type === 'OBSERVE' &&
      typeof edge.properties.key === 'string' &&
      edge.properties.key.length > 0
    ) {
      const key = edge.properties.key;
      const map = this._maps.get(path);
      const value = deserializeCrdtValue(
        this._resolveEdgeProperty(edge, 'value')
      );
      const hadPreviousValue = map?.has(key) ?? false;
      const previousValue = map?.applyRemoteSet(key, value);
      return hadPreviousValue
        ? {
            type: 'set',
            path,
            key,
            value,
            previousValue,
            origin,
          }
        : {
            type: 'set',
            path,
            key,
            value,
            origin,
          };
    }

    if (sourceId === `map_${path}` && edge.type === 'FORK') {
      const key = edge.properties.key;
      if (!key) {
        return null;
      }
      const map = this._maps.get(path);
      if (edge.properties.op === 'delete') {
        const hadPreviousValue = map?.has(key) ?? false;
        const previousValue = map?.applyRemoteDelete(key);
        return hadPreviousValue
          ? {
              type: 'delete',
              path,
              key,
              previousValue,
              origin,
            }
          : {
              type: 'delete',
              path,
              key,
              origin,
            };
      }

      return null;
    }

    if (sourceId === `arr_${path}` && edge.type === 'FORK') {
      const array = this._arrays.get(path);
      if (edge.properties.op === 'push') {
        const value = deserializeCrdtValue(
          this._resolveEdgeProperty(edge, 'value')
        );
        array?.applyRemotePush(value);
        return {
          type: 'insert',
          path,
          value,
          origin,
        };
      }

      if (edge.properties.op === 'delete') {
        const index = Number(edge.properties.pos);
        const length = Number(edge.properties.len);
        if (Number.isFinite(index) && Number.isFinite(length)) {
          array?.applyRemoteDelete(index, length);
          return {
            type: 'delete',
            path,
            origin,
          };
        }
      }

      return null;
    }

    if (sourceId === `text_${path}` && edge.type === 'FORK') {
      const text = this._texts.get(path);
      if (edge.properties.op === 'insert') {
        const index = Number(edge.properties.pos);
        const insertedText = edge.properties.text ?? '';
        if (Number.isFinite(index)) {
          text?.applyRemoteInsert(index, insertedText);
          return {
            type: 'insert',
            path,
            value: insertedText,
            origin,
          };
        }
      }

      if (edge.properties.op === 'delete') {
        const index = Number(edge.properties.pos);
        const length = Number(edge.properties.len);
        if (Number.isFinite(index) && Number.isFinite(length)) {
          text?.applyRemoteDelete(index, length);
          return {
            type: 'delete',
            path,
            origin,
          };
        }
      }

      return null;
    }

    if (sourceId === `ctr_${path}` && edge.type === 'FOLD') {
      const delta = Number(edge.properties.delta);
      if (!Number.isFinite(delta)) {
        return null;
      }
      this._counters.get(path)?.applyRemoteIncrement(delta);
      return {
        type: 'set',
        path,
        value: delta,
        origin,
      };
    }

    return null;
  }

  private _replayMapState(name: string, map: QMap<unknown>): void {
    for (const edge of this._edges) {
      if (
        edge.type === 'OBSERVE' &&
        edge.properties.path === name &&
        typeof edge.properties.key === 'string' &&
        edge.properties.key.length > 0
      ) {
        map.applyRemoteSet(
          edge.properties.key,
          deserializeCrdtValue(this._resolveEdgeProperty(edge, 'value'))
        );
        continue;
      }

      if (edge.sourceIds[0] !== `map_${name}` || edge.type !== 'FORK') {
        continue;
      }
      const key = edge.properties.key;
      if (!key) {
        continue;
      }
      if (edge.properties.op === 'delete') {
        map.applyRemoteDelete(key);
      }
    }
  }

  private _replayArrayState(name: string, array: QArray<unknown>): void {
    for (const edge of this._edges) {
      if (edge.sourceIds[0] !== `arr_${name}` || edge.type !== 'FORK') {
        continue;
      }
      if (edge.properties.op === 'push') {
        array.applyRemotePush(
          deserializeCrdtValue(this._resolveEdgeProperty(edge, 'value'))
        );
        continue;
      }
      if (edge.properties.op !== 'delete') {
        continue;
      }
      const index = Number(edge.properties.pos);
      const length = Number(edge.properties.len);
      if (Number.isFinite(index) && Number.isFinite(length)) {
        array.applyRemoteDelete(index, length);
      }
    }
  }

  private _replayTextState(name: string, text: QText): void {
    for (const edge of this._edges) {
      if (edge.sourceIds[0] !== `text_${name}` || edge.type !== 'FORK') {
        continue;
      }
      if (edge.properties.op === 'insert') {
        const index = Number(edge.properties.pos);
        const insertedText = edge.properties.text ?? '';
        if (Number.isFinite(index)) {
          text.applyRemoteInsert(index, insertedText);
        }
        continue;
      }
      if (edge.properties.op !== 'delete') {
        continue;
      }
      const index = Number(edge.properties.pos);
      const length = Number(edge.properties.len);
      if (Number.isFinite(index) && Number.isFinite(length)) {
        text.applyRemoteDelete(index, length);
      }
    }
  }

  private _replayCounterState(name: string, counter: QCounter): void {
    for (const edge of this._edges) {
      if (edge.sourceIds[0] !== `ctr_${name}` || edge.type !== 'FOLD') {
        continue;
      }
      const delta = Number(edge.properties.delta);
      if (Number.isFinite(delta)) {
        counter.applyRemoteIncrement(delta);
      }
    }
  }

  private _resolveEdgeProperty(
    edge: QDocDeltaEdge,
    propertyName: string
  ): string | undefined {
    const directValue = edge.properties[propertyName];
    if (typeof directValue === 'string') {
      return directValue;
    }

    const targetId = edge.targetIds[0];
    if (!targetId) {
      return undefined;
    }

    const targetNode = this._nodes.get(targetId);
    const nodeValue = targetNode?.properties[propertyName];
    return typeof nodeValue === 'string' ? nodeValue : undefined;
  }

  // ── Observability ──────────────────────────────────────────────────────

  /**
   * Listen for document updates (replaces Y.Doc.on('update')).
   */
  on(event: 'update', handler: QDocUpdateHandler): void;
  on(
    event: string,
    handler: QDocUpdateHandler | ((...args: unknown[]) => void)
  ): void {
    if (event === 'update') {
      this._updateHandlers.add(handler as QDocUpdateHandler);
    }
  }

  off(event: 'update', handler: QDocUpdateHandler): void;
  off(
    event: string,
    handler: QDocUpdateHandler | ((...args: unknown[]) => void)
  ): void {
    if (event === 'update') {
      this._updateHandlers.delete(handler as QDocUpdateHandler);
    }
  }

  /**
   * Observe changes at a specific path.
   */
  observe(path: string, handler: QDocObserveHandler): void {
    let handlers = this._observeHandlers.get(path);
    if (!handlers) {
      handlers = new Set();
      this._observeHandlers.set(path, handlers);
    }
    handlers.add(handler);
  }

  unobserve(path: string, handler: QDocObserveHandler): void {
    this._observeHandlers.get(path)?.delete(handler);
  }

  // ── Presence (SLIVER — never collapses) ─────────────────────────────

  /**
   * Set local presence state (replaces awareness.setLocalState).
   * Presence uses SLIVER — it never collapses, cursors coexist.
   */
  setPresence(state: Record<string, unknown>): void {
    this._presence.set(this._replicaId, state);
    this._notifyPresence();
  }

  /**
   * Get all presence states (replaces awareness.getStates).
   */
  getPresenceStates(): Map<string, Record<string, unknown>> {
    return new Map(this._presence);
  }

  /**
   * Apply remote presence (SLIVER — append, never merge).
   */
  applyPresence(replicaId: string, state: Record<string, unknown>): void {
    this._presence.set(replicaId, state);
    this._notifyPresence();
  }

  removePresence(replicaId: string): void {
    this._presence.delete(replicaId);
    this._notifyPresence();
  }

  onPresenceChange(
    handler: (states: Map<string, Record<string, unknown>>) => void
  ): void {
    this._presenceHandlers.add(handler);
  }

  offPresenceChange(
    handler: (states: Map<string, Record<string, unknown>>) => void
  ): void {
    this._presenceHandlers.delete(handler);
  }

  private _notifyPresence(): void {
    for (const handler of this._presenceHandlers) {
      try {
        handler(this.getPresenceStates());
      } catch {
        /* */
      }
    }
  }

  // ── Topology Metrics ──────────────────────────────────────────────────

  get nodeCount(): number {
    return this._nodes.size;
  }
  get edgeCount(): number {
    return this._edges.length;
  }

  /**
   * Get the full topology as a GG source string (for model checking).
   */
  toGG(): string {
    const lines: string[] = [];
    for (const node of this._nodes.values()) {
      const label = node.labels.length > 0 ? `: ${node.labels[0]}` : '';
      const props = Object.entries(node.properties);
      const propsStr =
        props.length > 0
          ? ` { ${props.map(([k, v]) => `${k}: '${v}'`).join(', ')} }`
          : '';
      lines.push(`(${node.id}${label}${propsStr})`);
    }
    for (const edge of this._edges) {
      const sources = edge.sourceIds.join(' | ');
      const targets = edge.targetIds.join(' | ');
      const props = Object.entries(edge.properties);
      const propsStr =
        props.length > 0
          ? ` { ${props.map(([k, v]) => `${k}: '${v}'`).join(', ')} }`
          : '';
      lines.push(`(${sources})-[:${edge.type}${propsStr}]->(${targets})`);
    }
    return lines.join('\n');
  }
}

// ── QMap — Replaces Y.Map ───────────────────────────────────────────────

export class QMap<T = unknown> {
  private readonly _doc: QDoc;
  private readonly _name: string;
  private readonly _strategy: GgCollapseStrategy;
  private readonly _data: Map<string, T> = new Map();
  private readonly _observeHandlers: Map<
    (event: {
      keysChanged: Set<string>;
      changes: {
        keys: Map<
          string,
          { action: 'add' | 'update' | 'delete'; oldValue?: T }
        >;
      };
    }) => void,
    QDocObserveHandler
  > = new Map();
  private _branchCounter = 0;

  constructor(
    doc: QDoc,
    name: string,
    strategy: GgCollapseStrategy,
    initializeTopology = true
  ) {
    this._doc = doc;
    this._name = name;
    this._strategy = strategy;
    if (initializeTopology) {
      // Create the map root node
      doc._appendNode({
        id: `map_${name}`,
        labels: ['QMap'],
        properties: { name, strategy },
      });
      doc._appendEdge({
        sourceIds: ['root'],
        targetIds: [`map_${name}`],
        type: 'PROCESS',
        properties: { path: name },
      });
    }
  }

  set(key: string, value: T): void {
    const hadPreviousValue = this._data.has(key);
    const previousValue = this._data.get(key);
    const branchId = `map_${this._name}_${key}_${this._branchCounter++}`;
    const valueStr = serializeCrdtValue(value);

    // FORK a write branch
    this._doc._appendNode({
      id: branchId,
      labels: ['Write'],
      properties: { key, value: valueStr, ts: String(this._doc.clock) },
    });
    this._doc._appendEdge({
      sourceIds: [`map_${this._name}`],
      targetIds: [branchId],
      type: 'FORK',
      properties: { path: this._name, key, op: 'set' },
    });

    // Immediately OBSERVE (eager collapse for local reads)
    const observeId = `map_${this._name}_${key}_obs_${this._branchCounter}`;
    this._doc._appendNode({
      id: observeId,
      labels: ['Observed'],
      properties: { key, value: valueStr },
    });
    this._doc._appendEdge({
      sourceIds: [branchId],
      targetIds: [observeId],
      type: 'OBSERVE',
      properties: {
        strategy: this._strategy,
        path: this._name,
        key,
        value: valueStr,
      },
    });

    this._data.set(key, value);
    this._doc._emitObserve(
      this._name,
      hadPreviousValue
        ? {
            type: 'set',
            path: this._name,
            key,
            value,
            previousValue,
            origin: 'local',
          }
        : {
            type: 'set',
            path: this._name,
            key,
            value,
            origin: 'local',
          }
    );
    this._doc._notifyUpdate('local');
  }

  get(key: string): T | undefined {
    return this._data.get(key);
  }

  has(key: string): boolean {
    return this._data.has(key);
  }

  applyRemoteSet(key: string, value: unknown): T | undefined {
    const previousValue = this._data.get(key);
    this._data.set(key, value as T);
    return previousValue;
  }

  delete(key: string): boolean {
    const hadPreviousValue = this._data.has(key);
    const previousValue = this._data.get(key);
    if (!this._data.delete(key)) {
      return false;
    }
    const branchId = `map_${this._name}_${key}_del_${this._branchCounter++}`;
    this._doc._appendNode({
      id: branchId,
      labels: ['Delete'],
      properties: { key },
    });
    this._doc._appendEdge({
      sourceIds: [`map_${this._name}`],
      targetIds: [branchId],
      type: 'FORK',
      properties: { path: this._name, key, op: 'delete' },
    });
    this._doc._emitObserve(
      this._name,
      hadPreviousValue
        ? {
            type: 'delete',
            path: this._name,
            key,
            previousValue,
            origin: 'local',
          }
        : {
            type: 'delete',
            path: this._name,
            key,
            origin: 'local',
          }
    );
    this._doc._notifyUpdate('local');
    return true;
  }

  applyRemoteDelete(key: string): T | undefined {
    const previousValue = this._data.get(key);
    this._data.delete(key);
    return previousValue;
  }

  toJSON(): Record<string, T> {
    const result: Record<string, T> = {} as Record<string, T>;
    for (const [key, value] of this._data) {
      result[key] = value;
    }
    return result;
  }

  get size(): number {
    return this._data.size;
  }

  forEach(fn: (value: T, key: string, map: QMap<T>) => void): void {
    this._data.forEach((value, key) => fn(value, key, this));
  }

  entries(): IterableIterator<[string, T]> {
    return this._data.entries();
  }

  keys(): IterableIterator<string> {
    return this._data.keys();
  }

  values(): IterableIterator<T> {
    return this._data.values();
  }

  // Y.Map compat surface used by downstream apps.
  get doc(): QDoc {
    return this._doc;
  }

  observe(handler: (event: QMapEvent<T>) => void): void {
    if (this._observeHandlers.has(handler)) {
      return;
    }
    const wrapped: QDocObserveHandler = (event) => {
      if (!event.key) {
        return;
      }
      const hasPreviousValue = Object.prototype.hasOwnProperty.call(
        event,
        'previousValue'
      );
      const action: 'add' | 'update' | 'delete' =
        event.type === 'delete'
          ? 'delete'
          : hasPreviousValue
          ? 'update'
          : 'add';
      const change: QMapKeyChange<T> =
        action === 'add'
          ? { action }
          : {
              action,
              ...(event.previousValue !== undefined && {
                oldValue: event.previousValue as T,
              }),
            };
      handler({
        keysChanged: new Set([event.key]),
        changes: {
          keys: new Map([[event.key, change]]),
        },
      });
    };
    this._observeHandlers.set(handler, wrapped);
    this._doc.observe(this._name, wrapped);
  }

  unobserve(handler: (event: QMapEvent<T>) => void): void {
    const wrapped = this._observeHandlers.get(handler);
    if (!wrapped) {
      return;
    }
    this._doc.unobserve(this._name, wrapped);
    this._observeHandlers.delete(handler);
  }
}

export class QXmlFragment {
  private readonly _doc: QDoc;
  private readonly _name: string;
  private readonly _children: unknown[] = [];

  constructor(doc: QDoc, name: string) {
    this._doc = doc;
    this._name = name;
  }

  insert(index: number, content: unknown[]): void {
    this._children.splice(index, 0, ...content);
    this._doc._notifyUpdate('local');
  }

  delete(index: number, length: number): void {
    this._children.splice(index, length);
    this._doc._notifyUpdate('local');
  }

  get(index: number): unknown {
    return this._children[index];
  }

  get length(): number {
    return this._children.length;
  }

  toArray(): unknown[] {
    return [...this._children];
  }

  toString(): string {
    return this._children.map((child) => String(child)).join('');
  }

  toJSON(): unknown[] {
    return [...this._children];
  }

  observe(handler: QDocObserveHandler): void {
    this._doc.observe(this._name, handler);
  }

  unobserve(handler: QDocObserveHandler): void {
    this._doc.unobserve(this._name, handler);
  }
}

// ── QArray — Replaces Y.Array ───────────────────────────────────────────

export class QArray<T = unknown> {
  private readonly _doc: QDoc;
  private readonly _name: string;
  private readonly _data: T[] = [];
  private _branchCounter = 0;

  constructor(
    doc: QDoc,
    name: string,
    strategy: GgCollapseStrategy,
    initializeTopology = true
  ) {
    this._doc = doc;
    this._name = name;
    if (initializeTopology) {
      doc._appendNode({
        id: `arr_${name}`,
        labels: ['QArray'],
        properties: { name, strategy },
      });
      doc._appendEdge({
        sourceIds: ['root'],
        targetIds: [`arr_${name}`],
        type: 'PROCESS',
        properties: { path: name },
      });
    }
  }

  push(items: T[]): void {
    for (const item of items) {
      const branchId = `arr_${this._name}_push_${this._branchCounter++}`;
      const valueStr = serializeCrdtValue(item);

      this._doc._appendNode({
        id: branchId,
        labels: ['Insert'],
        properties: { pos: String(this._data.length), value: valueStr },
      });
      this._doc._appendEdge({
        sourceIds: [`arr_${this._name}`],
        targetIds: [branchId],
        type: 'FORK',
        properties: { path: this._name, op: 'push', value: valueStr },
      });

      this._data.push(item);
    }
    this._doc._notifyUpdate('local');
  }

  delete(index: number, length = 1): void {
    this._data.splice(index, length);
    const branchId = `arr_${this._name}_del_${this._branchCounter++}`;
    this._doc._appendNode({
      id: branchId,
      labels: ['Delete'],
      properties: { pos: String(index), len: String(length) },
    });
    this._doc._appendEdge({
      sourceIds: [`arr_${this._name}`],
      targetIds: [branchId],
      type: 'FORK',
      properties: {
        path: this._name,
        op: 'delete',
        pos: String(index),
        len: String(length),
      },
    });
    this._doc._notifyUpdate('local');
  }

  get(index: number): T | undefined {
    return this._data[index];
  }

  get length(): number {
    return this._data.length;
  }

  toArray(): T[] {
    return [...this._data];
  }

  toJSON(): T[] {
    return this.toArray();
  }

  applyRemotePush(item: unknown): void {
    this._data.push(item as T);
  }

  applyRemoteDelete(index: number, length = 1): void {
    this._data.splice(index, length);
  }

  forEach(fn: (item: T, index: number) => void): void {
    this._data.forEach(fn);
  }

  observe(
    handler: (event: { changes: { delta: Array<{ insert?: T[] }> } }) => void
  ): void {
    this._doc.observe(this._name, (event) => {
      if (event.value !== undefined) {
        handler({
          changes: {
            delta: [{ insert: [event.value as T] }],
          },
        });
      }
    });
  }

  unobserve(_handler: (event: unknown) => void): void {
    // Handled via doc.unobserve
  }
}

// ── QText — Replaces Y.Text ────────────────────────────────────────────

export class QText {
  private readonly _doc: QDoc;
  private readonly _name: string;
  private _content = '';
  private _branchCounter = 0;

  constructor(doc: QDoc, name: string, initializeTopology = true) {
    this._doc = doc;
    this._name = name;
    if (initializeTopology) {
      doc._appendNode({
        id: `text_${name}`,
        labels: ['QText'],
        properties: { name },
      });
      doc._appendEdge({
        sourceIds: ['root'],
        targetIds: [`text_${name}`],
        type: 'PROCESS',
        properties: { path: name },
      });
    }
  }

  insert(index: number, text: string): void {
    this._content =
      this._content.slice(0, index) + text + this._content.slice(index);
    const branchId = `text_${this._name}_ins_${this._branchCounter++}`;
    this._doc._appendNode({
      id: branchId,
      labels: ['Insert'],
      properties: { pos: String(index), text },
    });
    this._doc._appendEdge({
      sourceIds: [`text_${this._name}`],
      targetIds: [branchId],
      type: 'FORK',
      properties: { path: this._name, op: 'insert', pos: String(index), text },
    });
    this._doc._notifyUpdate('local');
  }

  delete(index: number, length: number): void {
    this._content =
      this._content.slice(0, index) + this._content.slice(index + length);
    const branchId = `text_${this._name}_del_${this._branchCounter++}`;
    this._doc._appendNode({
      id: branchId,
      labels: ['Delete'],
      properties: { pos: String(index), len: String(length) },
    });
    this._doc._appendEdge({
      sourceIds: [`text_${this._name}`],
      targetIds: [branchId],
      type: 'FORK',
      properties: {
        path: this._name,
        op: 'delete',
        pos: String(index),
        len: String(length),
      },
    });
    this._doc._notifyUpdate('local');
  }

  toString(): string {
    return this._content;
  }

  get length(): number {
    return this._content.length;
  }

  toJSON(): string {
    return this._content;
  }

  applyRemoteInsert(index: number, text: string): void {
    this._content =
      this._content.slice(0, index) + text + this._content.slice(index);
  }

  applyRemoteDelete(index: number, length: number): void {
    this._content =
      this._content.slice(0, index) + this._content.slice(index + length);
  }
}

// ── QCounter — Replaces shared counter patterns ────────────────────────

export class QCounter {
  private readonly _doc: QDoc;
  private readonly _name: string;
  private _value = 0;
  private _branchCounter = 0;

  constructor(doc: QDoc, name: string, initializeTopology = true) {
    this._doc = doc;
    this._name = name;
    if (initializeTopology) {
      doc._appendNode({
        id: `ctr_${name}`,
        labels: ['QCounter'],
        properties: { name, initial: '0' },
      });
      doc._appendEdge({
        sourceIds: ['root'],
        targetIds: [`ctr_${name}`],
        type: 'PROCESS',
        properties: { path: name },
      });
    }
  }

  increment(delta = 1): void {
    this._value += delta;
    const branchId = `ctr_${this._name}_inc_${this._branchCounter++}`;
    this._doc._appendNode({
      id: branchId,
      labels: ['Increment'],
      properties: { delta: String(delta) },
    });
    // FOLD not OBSERVE — counters are commutative
    this._doc._appendEdge({
      sourceIds: [`ctr_${this._name}`],
      targetIds: [branchId],
      type: 'FOLD',
      properties: {
        strategy: 'fold-sum',
        path: this._name,
        delta: String(delta),
      },
    });
    this._doc._notifyUpdate('local');
  }

  decrement(delta = 1): void {
    this.increment(-delta);
  }

  get value(): number {
    return this._value;
  }

  toJSON(): number {
    return this._value;
  }

  applyRemoteIncrement(delta: number): void {
    this._value += delta;
  }
}
