// @a0n/.ignored_gnosis/src/crdt/yjs-compat.ts
'use aeon';

/**
 * Yjs compatibility surface backed by gnosis-owned code.
 *
 * This module exists for broad migration: existing Yjs call sites can
 * switch imports from `yjs` to `@a0n/gnosis` without pulling
 * external Yjs at runtime.
 */

type UpdateHandler = (update: Uint8Array, origin: unknown) => void;

class MockYText {
  _content = '';
  _observers: Array<() => void> = [];
  _doc: Doc | null = null;

  insert(index: number, text: string): void {
    this._content =
      this._content.slice(0, index) + text + this._content.slice(index);
    this._notifyDoc();
  }

  delete(index: number, length: number): void {
    this._content =
      this._content.slice(0, index) + this._content.slice(index + length);
    this._notifyDoc();
  }

  private _notifyDoc(): void {
    if (!this._doc) {
      return;
    }
    const handlers = this._doc._listeners.get('update') || [];
    for (const handler of handlers) {
      handler(new Uint8Array(0), null);
    }
  }

  toString(): string {
    return this._content;
  }

  toJSON(): string {
    return this._content;
  }

  observe(fn: () => void): void {
    this._observers.push(fn);
  }

  unobserve(): void {
    // no-op
  }
}

class MockYMap {
  _map = new Map<string, unknown>();
  _observers: Array<(event: unknown) => void> = [];
  _doc: Doc | null = null;

  get doc(): Doc | null {
    return this._doc;
  }

  set(k: string, v: unknown): void {
    const action = this._map.has(k) ? 'update' : 'add';
    this._map.set(k, v);
    this._fireObservers(k, action);
  }

  get(k: string): unknown {
    return this._map.get(k);
  }

  has(k: string): boolean {
    return this._map.has(k);
  }

  delete(k: string): boolean {
    const had = this._map.has(k);
    const result = this._map.delete(k);
    if (had) {
      this._fireObservers(k, 'delete');
    }
    return result;
  }

  toJSON(): Record<string, unknown> {
    return Object.fromEntries(this._map);
  }

  forEach(fn: (value: unknown, key: string) => void): void {
    this._map.forEach((value, key) => fn(value, key));
  }

  entries(): IterableIterator<[string, unknown]> {
    return this._map.entries();
  }

  keys(): IterableIterator<string> {
    return this._map.keys();
  }

  values(): IterableIterator<unknown> {
    return this._map.values();
  }

  [Symbol.iterator](): IterableIterator<[string, unknown]> {
    return this._map[Symbol.iterator]();
  }

  get size(): number {
    return this._map.size;
  }

  observe(fn: (event: unknown) => void): void {
    this._observers.push(fn);
  }

  unobserve(fn: (event: unknown) => void): void {
    const index = this._observers.indexOf(fn);
    if (index >= 0) {
      this._observers.splice(index, 1);
    }
  }

  private _fireObservers(key: string, action: string): void {
    const changes = new Map([[key, { action }]]);
    const event = { changes: { keys: changes } };
    for (const observer of this._observers) {
      observer(event);
    }
  }
}

class MockYArray {
  _arr: unknown[] = [];
  _observers: Array<
    (event: { changes: { delta: Array<{ insert?: unknown[] }> } }) => void
  > = [];

  push(items: unknown[]): void {
    this._arr.push(...items);
    this._emitInsert(items);
  }

  insert(index: number, content: unknown[]): void {
    this._arr.splice(index, 0, ...content);
    this._emitInsert(content);
  }

  delete(index: number, length: number): void {
    this._arr.splice(index, length);
  }

  get(index: number): unknown {
    return this._arr[index];
  }

  toArray(): unknown[] {
    return [...this._arr];
  }

  toJSON(): unknown[] {
    return [...this._arr];
  }

  get length(): number {
    return this._arr.length;
  }

  forEach(fn: (value: unknown, index: number, array: unknown[]) => void): void {
    this._arr.forEach(fn);
  }

  map<T>(fn: (value: unknown, index: number, array: unknown[]) => T): T[] {
    return this._arr.map(fn);
  }

  observe(
    fn: (event: { changes: { delta: Array<{ insert?: unknown[] }> } }) => void
  ): void {
    this._observers.push(fn);
  }

  unobserve(
    fn: (event: { changes: { delta: Array<{ insert?: unknown[] }> } }) => void
  ): void {
    const index = this._observers.indexOf(fn);
    if (index >= 0) {
      this._observers.splice(index, 1);
    }
  }

  private _emitInsert(insert: unknown[]): void {
    if (insert.length === 0) {
      return;
    }
    const event = { changes: { delta: [{ insert }] } };
    for (const observer of this._observers) {
      observer(event);
    }
  }
}

class MockXmlElement {
  readonly nodeName: string;
  private readonly attrs = new Map<string, string>();
  private textContent = '';

  constructor(nodeName: string = 'div') {
    this.nodeName = nodeName;
  }

  setAttribute(key: string, value: string): void {
    this.attrs.set(key, value);
  }

  getAttribute(key: string): string | null {
    return this.attrs.get(key) ?? null;
  }

  insert(index: number, text: string): void {
    this.textContent =
      this.textContent.slice(0, index) + text + this.textContent.slice(index);
  }

  delete(index: number, length: number): void {
    this.textContent =
      this.textContent.slice(0, index) + this.textContent.slice(index + length);
  }

  toString(): string {
    return this.textContent;
  }
}

class MockXmlFragment {
  _children: unknown[] = [];
  _observers: Array<(event: unknown) => void> = [];

  insert(index: number, content: unknown[]): void {
    this._children.splice(index, 0, ...content);
  }

  delete(index: number, length: number): void {
    this._children.splice(index, length);
  }

  get(index: number): unknown {
    return this._children[index];
  }

  get length(): number {
    return this._children.length;
  }

  toArray(): unknown[] {
    return [...this._children];
  }

  toString(): string {
    return this._children.map((child) => String(child)).join('');
  }

  toJSON(): unknown[] {
    return [...this._children];
  }

  observe(fn: (event: unknown) => void): void {
    this._observers.push(fn);
  }

  unobserve(fn: (event: unknown) => void): void {
    const index = this._observers.indexOf(fn);
    if (index >= 0) {
      this._observers.splice(index, 1);
    }
  }
}

export class Doc {
  clientID = Math.floor(Math.random() * 1e9);
  gc = true;
  _texts = new Map<string, MockYText>();
  _maps = new Map<string, MockYMap>();
  _arrays = new Map<string, MockYArray>();
  _xmlFragments = new Map<string, MockXmlFragment>();
  _listeners = new Map<string, UpdateHandler[]>();

  getText(name: string): MockYText {
    let text = this._texts.get(name);
    if (!text) {
      text = new MockYText();
      text._doc = this;
      this._texts.set(name, text);
    }
    return text;
  }

  getMap(name: string): MockYMap {
    let map = this._maps.get(name);
    if (!map) {
      map = new MockYMap();
      map._doc = this;
      this._maps.set(name, map);
    }
    return map;
  }

  getArray(name: string): MockYArray {
    let array = this._arrays.get(name);
    if (!array) {
      array = new MockYArray();
      this._arrays.set(name, array);
    }
    return array;
  }

  getXmlFragment(name: string): MockXmlFragment {
    let fragment = this._xmlFragments.get(name);
    if (!fragment) {
      fragment = new MockXmlFragment();
      this._xmlFragments.set(name, fragment);
    }
    return fragment;
  }

  transact(fn: () => void, origin?: unknown): void {
    fn();
    const handlers = this._listeners.get('update') || [];
    for (const handler of handlers) {
      handler(new Uint8Array(0), origin);
    }
  }

  on(event: string, fn: UpdateHandler): void {
    if (!this._listeners.has(event)) {
      this._listeners.set(event, []);
    }
    this._listeners.get(event)?.push(fn);
  }

  off(event: string, fn: UpdateHandler): void {
    const handlers = this._listeners.get(event);
    if (!handlers) {
      return;
    }
    const index = handlers.indexOf(fn);
    if (index >= 0) {
      handlers.splice(index, 1);
    }
  }

  destroy(): void {
    this._listeners.clear();
  }
}

/**
 * Handles the gnosis encode State As Update workflow.
 */
export function encodeStateAsUpdate(
  doc: Doc,
  _stateVector?: Uint8Array
): Uint8Array {
  const snapshot: Record<string, Record<string, unknown>> = {
    texts: {},
    maps: {},
    arrays: {},
  };

  doc._texts.forEach((text, key) => {
    snapshot.texts![key] = text._content;
  });
  doc._maps.forEach((map, key) => {
    snapshot.maps![key] = Object.fromEntries(map._map);
  });
  doc._arrays.forEach((array, key) => {
    snapshot.arrays![key] = [...array._arr];
  });

  return new TextEncoder().encode(JSON.stringify(snapshot));
}

/**
 * Handles the gnosis apply Update workflow.
 */
export function applyUpdate(doc: Doc, update: Uint8Array): void {
  try {
    const raw = new TextDecoder('utf-8', {
      fatal: true,
      ignoreBOM: false,
    }).decode(update);
    const snapshot = JSON.parse(raw) as {
      texts?: Record<string, string>;
      maps?: Record<string, Record<string, unknown>>;
      arrays?: Record<string, unknown[]>;
    };

    if (snapshot.texts) {
      for (const [key, value] of Object.entries(snapshot.texts)) {
        const text = doc.getText(key);
        text._content = value;
      }
    }

    if (snapshot.maps) {
      for (const [key, value] of Object.entries(snapshot.maps)) {
        const map = doc.getMap(key);
        for (const [entryKey, entryValue] of Object.entries(value)) {
          map.set(entryKey, entryValue);
        }
      }
    }

    if (snapshot.arrays) {
      for (const [key, value] of Object.entries(snapshot.arrays)) {
        const array = doc.getArray(key);
        array._arr = [...value];
      }
    }
  } catch {
    // Ignore invalid payloads.
  }
}

/**
 * Handles the gnosis transact workflow.
 */
export function transact(doc: Doc, fn: () => void, origin?: unknown): void {
  doc.transact(fn, origin);
}

/**
 * Handles the gnosis encode State Vector workflow.
 */
export const encodeStateVector = (): Uint8Array => new Uint8Array(0);
/**
 * Handles the gnosis diff Update workflow.
 */
export const diffUpdate = (update: Uint8Array, _sv: Uint8Array): Uint8Array =>
  update;
/**
 * Handles the gnosis merge Updates workflow.
 */
export const mergeUpdates = (updates: Uint8Array[]): Uint8Array =>
  updates[0] ?? new Uint8Array(0);

class MockUndoManager {
  private readonly scope: MockYText;
  private readonly trackedOrigins: Set<unknown>;
  private readonly undoStack: Array<{ content: string }> = [];
  private readonly redoStack: Array<{ content: string }> = [];
  private readonly doc: Doc | null;

  constructor(scope: MockYText, options?: { trackedOrigins?: Set<unknown> }) {
    this.scope = scope;
    this.trackedOrigins = options?.trackedOrigins ?? new Set();
    this.doc = scope._doc;

    if (this.doc) {
      this.doc.on('update', (_update, origin) => {
        if (this.trackedOrigins.has(origin)) {
          this.undoStack.push({ content: this.scope.toString() });
          this.redoStack.length = 0;
        }
      });
    }
  }

  undo(): void {
    const item = this.undoStack.pop();
    if (!item) {
      return;
    }
    this.redoStack.push({ content: this.scope.toString() });
    this.scope.delete(0, this.scope._content.length);
    if (this.undoStack.length > 0) {
      const previous = this.undoStack[this.undoStack.length - 1];
      if (previous) this.scope.insert(0, previous.content);
    }
  }

  redo(): void {
    const item = this.redoStack.pop();
    if (!item) {
      return;
    }
    this.undoStack.push({ content: this.scope.toString() });
    this.scope.delete(0, this.scope._content.length);
    this.scope.insert(0, item.content);
  }

  destroy(): void {
    // no-op
  }

  clear(): void {
    this.undoStack.length = 0;
    this.redoStack.length = 0;
  }
}

export {
  MockUndoManager as UndoManager,
  MockYMap as Map,
  MockYArray as Array,
  MockYText as Text,
  MockXmlFragment as XmlFragment,
  MockXmlElement as XmlElement,
};
