// @a0n/.ignored_gnosis/src/capabilities/inference.ts
import { parseGgProgram } from '@a0n/aeon-logic';
import { BettyCompiler } from '../betty/compiler';
import type { CapabilityRequirement, HostCapability } from './types';
import { lowerUfcsSource } from '../ufcs';

const DECLARED_CAPABILITY_FIELDS = [
  'capability',
  'capabilities',
  'effect',
  'effects',
  'require',
  'requires',
] as const;

const CAPABILITY_ALIASES: Record<string, HostCapability> = {
  tcp: 'net.tcp.client',
  'tcp-client': 'net.tcp.client',
  'tcp.client': 'net.tcp.client',
  'net.tcp.client': 'net.tcp.client',
  'tcp-server': 'net.tcp.server',
  'tcp.server': 'net.tcp.server',
  'net.tcp.server': 'net.tcp.server',
  udp: 'net.udp',
  dgram: 'net.udp',
  'udp-socket': 'net.udp',
  'net.udp': 'net.udp',
  fs: 'fs.local',
  'fs-local': 'fs.local',
  'fs.local': 'fs.local',
  'fs-durable': 'fs.durable',
  'fs.durable': 'fs.durable',
  disk: 'fs.durable',
  durable: 'fs.durable',
  ucan: 'auth.ucan',
  'auth.ucan': 'auth.ucan',
  zk: 'auth.zk',
  'auth.zk': 'auth.zk',
  custodial: 'auth.custodial',
  'auth.custodial': 'auth.custodial',
};

const UCAN_LABELS = new Set([
  'ucanidentity',
  'ucanissue',
  'ucanverify',
  'ucandelegate',
  'ucanrequire',
]);

const ZK_LABELS = new Set([
  'zkencrypt',
  'zkdecrypt',
  'zksyncenvelope',
  'zkmaterializeenvelope',
]);

const CUSTODIAL_LABELS = new Set(['custodialsigner']);

function normalizeToken(value: string): string {
  return value
    .trim()
    .toLowerCase()
    .replace(/^['"]|['"]$/g, '');
}

function parseCapabilityTokens(value: string): string[] {
  return value
    .split(/[\s,|]+/)
    .map((entry) => normalizeToken(entry))
    .filter((entry) => entry.length > 0);
}

function readStringProperty(
  properties: Record<string, string>,
  key: string
): string | null {
  const value = properties[key];
  if (typeof value !== 'string') {
    return null;
  }
  return value;
}

function addRequirement(
  requirements: CapabilityRequirement[],
  capability: HostCapability,
  nodeId: string,
  reason: string,
  source: CapabilityRequirement['source'],
  label?: string
): void {
  requirements.push({ capability, nodeId, reason, source, ...(label !== undefined && { label }) });
}

function inferDeclaredCapabilities(
  requirements: CapabilityRequirement[],
  nodeId: string,
  label: string | undefined,
  properties: Record<string, string>
): void {
  for (const field of DECLARED_CAPABILITY_FIELDS) {
    const raw = readStringProperty(properties, field);
    if (!raw) continue;

    for (const token of parseCapabilityTokens(raw)) {
      const capability = CAPABILITY_ALIASES[token];
      if (!capability) continue;
      addRequirement(
        requirements,
        capability,
        nodeId,
        `Declared via ${field}`,
        'declaration',
        label
      );
    }
  }
}

function inferTransportCapabilities(
  requirements: CapabilityRequirement[],
  nodeId: string,
  label: string | undefined,
  properties: Record<string, string>
): void {
  const transport = normalizeToken(
    properties.transport ?? properties.protocol ?? ''
  );
  const mode = normalizeToken(
    properties.mode ?? properties.role ?? properties.op ?? ''
  );

  if (transport === 'udp') {
    addRequirement(
      requirements,
      'net.udp',
      nodeId,
      'Transport/protocol indicates UDP networking',
      'inference',
      label
    );
    return;
  }

  if (transport === 'tcp') {
    const capability =
      mode.includes('server') ||
      mode.includes('listen') ||
      mode.includes('bind')
        ? 'net.tcp.server'
        : 'net.tcp.client';

    addRequirement(
      requirements,
      capability,
      nodeId,
      'Transport/protocol indicates TCP networking',
      'inference',
      label
    );
  }
}

function inferStorageCapabilities(
  requirements: CapabilityRequirement[],
  nodeId: string,
  label: string | undefined,
  properties: Record<string, string>
): void {
  const storage = normalizeToken(
    properties.storage ?? properties.persistence ?? ''
  );
  const durableFlag = normalizeToken(properties.durable ?? '');
  const op = normalizeToken(properties.op ?? '');

  if (storage === 'durable' || durableFlag === 'true') {
    addRequirement(
      requirements,
      'fs.durable',
      nodeId,
      'Node declares durable local persistence',
      'inference',
      label
    );
    return;
  }

  if (op === 'read_file' || op === 'write_file') {
    addRequirement(
      requirements,
      'fs.local',
      nodeId,
      `Node operation ${op} requires local filesystem access`,
      'inference',
      label
    );
  }
}

function inferLabelCapabilities(
  requirements: CapabilityRequirement[],
  nodeId: string,
  labels: readonly string[]
): void {
  for (const rawLabel of labels) {
    const label = normalizeToken(rawLabel);

    if (UCAN_LABELS.has(label)) {
      addRequirement(
        requirements,
        'auth.ucan',
        nodeId,
        `Label ${rawLabel} uses UCAN capability semantics`,
        'inference',
        rawLabel
      );
    }

    if (ZK_LABELS.has(label)) {
      addRequirement(
        requirements,
        'auth.zk',
        nodeId,
        `Label ${rawLabel} uses zero-knowledge encryption primitives`,
        'inference',
        rawLabel
      );
    }

    if (CUSTODIAL_LABELS.has(label)) {
      addRequirement(
        requirements,
        'auth.custodial',
        nodeId,
        `Label ${rawLabel} uses custodial signing contracts`,
        'inference',
        rawLabel
      );
    }

    if (label.includes('tcpserver') || label.includes('tcplistener')) {
      addRequirement(
        requirements,
        'net.tcp.server',
        nodeId,
        `Label ${rawLabel} implies TCP server usage`,
        'inference',
        rawLabel
      );
    } else if (label.includes('tcp')) {
      addRequirement(
        requirements,
        'net.tcp.client',
        nodeId,
        `Label ${rawLabel} implies TCP client usage`,
        'inference',
        rawLabel
      );
    }

    if (label.includes('udp') || label.includes('dgram')) {
      addRequirement(
        requirements,
        'net.udp',
        nodeId,
        `Label ${rawLabel} implies UDP socket usage`,
        'inference',
        rawLabel
      );
    }
  }
}

function dedupeRequirements(
  requirements: CapabilityRequirement[]
): CapabilityRequirement[] {
  const seen = new Set<string>();
  const unique: CapabilityRequirement[] = [];

  for (const requirement of requirements) {
    const key = `${requirement.capability}::${requirement.nodeId}::${requirement.reason}`;
    if (seen.has(key)) continue;
    seen.add(key);
    unique.push(requirement);
  }

  return unique;
}

interface CapabilityInferenceNode {
  id: string;
  labels: readonly string[];
  properties: Record<string, string>;
}

function parseCapabilityInferenceNodes(
  source: string
): CapabilityInferenceNode[] {
  const normalizedSource = lowerUfcsSource(source);

  try {
    const program = parseGgProgram(normalizedSource);
    return program.nodes.map((node: any) => ({
      id: node.id,
      labels: node.labels ?? [],
      properties: (node.properties ?? {}) as Record<string, string>,
    }));
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    if (!message.includes('No .gg topology edges were parsed.')) {
      throw error;
    }

    const compiler = new BettyCompiler();
    const { ast, diagnostics } = compiler.parse(normalizedSource);
    const errors = diagnostics
      .filter((diagnostic) => diagnostic.severity === 'error')
      .map((diagnostic) => diagnostic.message);

    if (errors.length > 0 || !ast) {
      throw new Error(errors.length > 0 ? errors.join('; ') : message);
    }

    return [...ast.nodes.values()].map((node) => ({
      id: node.id,
      labels: node.labels ?? [],
      properties: node.properties ?? {},
    }));
  }
}

/**
 * Handles the gnosis infer Capabilities From Gg Source workflow.
 */
export function inferCapabilitiesFromGgSource(
  source: string
): CapabilityRequirement[] {
  const requirements: CapabilityRequirement[] = [];
  const nodes = parseCapabilityInferenceNodes(source);

  for (const node of nodes) {
    const nodeId = node.id;
    const labels = node.labels ?? [];
    const properties = (node.properties ?? {}) as Record<string, string>;
    const primaryLabel = labels[0];

    inferDeclaredCapabilities(requirements, nodeId, primaryLabel, properties);
    inferTransportCapabilities(requirements, nodeId, primaryLabel, properties);
    inferStorageCapabilities(requirements, nodeId, primaryLabel, properties);
    inferLabelCapabilities(requirements, nodeId, labels);
  }

  return dedupeRequirements(requirements);
}
