// @a0n/.ignored_gnosis/src/capabilities/summary.ts
import type { CapabilityRequirement, HostCapability } from './types';

export interface CapabilityContractNodeSummary {
  nodeId: string;
  declared: HostCapability[];
  inferred: HostCapability[];
}

export interface CapabilityContractSummary {
  declared: HostCapability[];
  inferred: HostCapability[];
  required: HostCapability[];
  nodes: CapabilityContractNodeSummary[];
}

function uniqueSortedCapabilities(
  capabilities: Iterable<HostCapability>
): HostCapability[] {
  return [...new Set(capabilities)].sort();
}

/**
 * Handles the gnosis summarize Capability Requirements workflow.
 */
export function summarizeCapabilityRequirements(
  requirements: readonly CapabilityRequirement[]
): CapabilityContractSummary {
  const nodes = new Map<string, CapabilityContractNodeSummary>();

  for (const requirement of requirements) {
    const existing = nodes.get(requirement.nodeId) ?? {
      nodeId: requirement.nodeId,
      declared: [],
      inferred: [],
    };

    if (requirement.source === 'declaration') {
      existing.declared.push(requirement.capability);
    } else {
      existing.inferred.push(requirement.capability);
    }

    nodes.set(requirement.nodeId, existing);
  }

  const nodeSummaries = [...nodes.values()]
    .map((summary) => ({
      nodeId: summary.nodeId,
      declared: uniqueSortedCapabilities(summary.declared),
      inferred: uniqueSortedCapabilities(summary.inferred),
    }))
    .sort((left, right) => left.nodeId.localeCompare(right.nodeId));

  return {
    declared: uniqueSortedCapabilities(
      requirements
        .filter((requirement) => requirement.source === 'declaration')
        .map((requirement) => requirement.capability)
    ),
    inferred: uniqueSortedCapabilities(
      requirements
        .filter((requirement) => requirement.source === 'inference')
        .map((requirement) => requirement.capability)
    ),
    required: uniqueSortedCapabilities(
      requirements.map((requirement) => requirement.capability)
    ),
    nodes: nodeSummaries,
  };
}
