// @a0n/.ignored_gnosis/src/capabilities/types.ts
export type RuntimeTarget =
  | 'agnostic'
  | 'workers'
  | 'tauri-edge'
  | 'node'
  | 'bun'
  | 'gnode'
  | 'wasm-browser'
  | 'wasi';

export type HostCapability =
  | 'net.tcp.client'
  | 'net.tcp.server'
  | 'net.udp'
  | 'fs.local'
  | 'fs.durable'
  | 'auth.ucan'
  | 'auth.zk'
  | 'auth.custodial'
  | 'process.spawn'
  | 'process.docker';

export interface CapabilityRequirement {
  capability: HostCapability;
  nodeId: string;
  reason: string;
  source: 'declaration' | 'inference';
  label?: string;
}

export interface CapabilityIssue {
  capability: HostCapability;
  severity: 'error' | 'warning';
  message: string;
  target: RuntimeTarget;
  nodeId?: string;
}

export interface CapabilityProfile {
  target: RuntimeTarget;
  supported: ReadonlySet<HostCapability>;
  description: string;
}

export interface CapabilityValidationReport {
  target: RuntimeTarget;
  required: CapabilityRequirement[];
  requiredUnique: HostCapability[];
  issues: CapabilityIssue[];
  ok: boolean;
}
