// @a0n/gnosis/moonshine-tauri/scripts/mobile-bridge.mjs
#!/usr/bin/env node
import { randomUUID } from 'node:crypto';
import { once } from 'node:events';
import { createServer } from 'node:http';
import {
  createWriteStream,
  mkdirSync,
  readFileSync,
  realpathSync,
  renameSync,
  statSync,
  unlinkSync,
  writeFileSync,
} from 'node:fs';
import { homedir } from 'node:os';
import { dirname, join } from 'node:path';
import { spawn } from 'node:child_process';
import { fileURLToPath } from 'node:url';

const appDir = dirname(fileURLToPath(import.meta.url));
const gnosisRoot = join(appDir, '..', '..');
const openSourceRoot = join(gnosisRoot, '..');
const port = Number.parseInt(process.env.MOONSHINE_MOBILE_BRIDGE_PORT || '14231', 10);
const host = process.env.MOONSHINE_MOBILE_BRIDGE_HOST || '0.0.0.0';
const token = process.env.MOONSHINE_MOBILE_BRIDGE_TOKEN || 'moonshine-dev';
const guardHome = process.env.MONSTER_GUARD_HOME || join(homedir(), '.moonshine.d');
const sidecarDir = join(appDir, '..', 'src-tauri', 'binaries');
const modelCacheDir = process.env.MOONSHINE_MODEL_CACHE_DIR || join(homedir(), '.edgework', 'models');
const modelConfigPath = process.env.MOONSHINE_MODEL_CONFIG || join(guardHome, 'mobile-model.json');
const skymeshConfigPath = process.env.MOONSHINE_SKYMESH_CONFIG || join(guardHome, 'skymesh.json');
const skymeshStatusPath = process.env.MOONSHINE_SKYMESH_STATUS || join(guardHome, 'skymesh-status.json');

const modelManifest = [
  {
    id: 'qwen2.5-0.5b-instruct-q4',
    label: 'Qwen2.5 0.5B instruct Q4',
    envUrl: 'MOONSHINE_QWEN25_05B_KNOT_URL',
    cacheName: 'qwen2.5-0.5b-instruct.knot',
    url:
      'https://www-edgework-app.edgework.ai/api/v1/r2/distributed-inference/models/qwen2.5-0.5b-instruct.knot',
  },
  {
    id: 'qwen-coder-7b',
    label: 'Qwen Coder 7B',
    envUrl: 'MOONSHINE_QWEN_CODER_7B_KNOT_URL',
    cacheName: 'qwen-coder-7b.knot',
    url: 'https://www-edgework-app.edgework.ai/api/v1/r2/distributed-inference/models/qwen-coder-7b.knot',
  },
  {
    id: 'qwen2.5-72b-instruct',
    label: 'Qwen2.5 72B instruct',
    envUrl: 'MOONSHINE_QWEN25_72B_KNOT_URL',
    cacheName: 'qwen2.5-72b-instruct.knot',
    url:
      'https://www-edgework-app.edgework.ai/api/v1/r2/distributed-inference/models/qwen2.5-72b-instruct.knot',
  },
];

function resolveBin(envName, fallback) {
  const explicit = process.env[envName];
  if (explicit && statSync(explicit, { throwIfNoEntry: false })?.isFile()) {
    return realpathSync(explicit);
  }
  const local = join(openSourceRoot, '.bin', fallback);
  if (statSync(local, { throwIfNoEntry: false })?.isFile()) {
    return realpathSync(local);
  }
  return fallback;
}

function existingDir(path) {
  return statSync(path, { throwIfNoEntry: false })?.isDirectory() ? realpathSync(path) : null;
}

function toolPath() {
  const dirs = [
    sidecarDir,
    join(openSourceRoot, 'aeon', 'packages', 'wall'),
    join(openSourceRoot, '.bin'),
    join(gnosisRoot, '.bin'),
    join(gnosisRoot, 'bin'),
    join(gnosisRoot, 'polyglot', 'target', 'release'),
    join(gnosisRoot, 'distributed-inference', 'target', 'release'),
    join(openSourceRoot, 'x-gnosis', 'gnosis-uring', 'target', 'release'),
  ].map(existingDir).filter(Boolean);
  const inherited = process.env.MOONSHINE_EXTRA_PATH;
  return inherited ? [...dirs, inherited].join(':') : dirs.join(':');
}

const moonshineExtraPath = toolPath();

const guardBin = resolveBin('MONSTER_GUARD_BIN', 'monster-guard');
const moonshineBin = resolveBin('MOONSHINE_BIN', 'moonshine');
const skymeshBridgeScript = join(appDir, 'moonshine-skymesh-bridge.mjs');
let skymeshBridge = null;
const processes = new Map();

function readJson(path, fallback) {
  try {
    return JSON.parse(readFileSync(path, 'utf8'));
  } catch {
    return fallback;
  }
}

function writeJson(path, value) {
  mkdirSync(dirname(path), { recursive: true });
  writeFileSync(path, `${JSON.stringify(value, null, 2)}\n`);
}

function readModelConfig() {
  try {
    const parsed = JSON.parse(readFileSync(modelConfigPath, 'utf8'));
    return {
      activeModelId:
        typeof parsed.activeModelId === 'string' ? parsed.activeModelId : modelManifest[0].id,
      urlOverrides:
        parsed.urlOverrides && typeof parsed.urlOverrides === 'object' ? parsed.urlOverrides : {},
    };
  } catch {
    return { activeModelId: modelManifest[0].id, urlOverrides: {} };
  }
}

function writeModelConfig(config) {
  writeJson(modelConfigPath, config);
}

function readSkymeshConfig() {
  const parsed = readJson(skymeshConfigPath, {});
  return {
    enabled: parsed.enabled !== false,
  };
}

function writeSkymeshConfig(config) {
  writeJson(skymeshConfigPath, config);
}

function modelById(id) {
  return modelManifest.find((model) => model.id === id) || modelManifest[0];
}

function modelUrl(model, config = readModelConfig()) {
  const override = config.urlOverrides?.[model.id];
  if (typeof override === 'string' && override.trim()) return override.trim();
  const envUrl = model.envUrl ? process.env[model.envUrl] : null;
  return envUrl && envUrl.trim() ? envUrl.trim() : model.url;
}

function modelSnapshot(model, config = readModelConfig()) {
  const cachePath = join(modelCacheDir, model.cacheName);
  const stat = statSync(cachePath, { throwIfNoEntry: false });
  return {
    id: model.id,
    label: model.label,
    url: modelUrl(model, config),
    cache_path: cachePath,
    present: !!stat?.isFile(),
    size_bytes: stat?.isFile() ? stat.size : null,
    active: model.id === config.activeModelId,
  };
}

function modelStatus() {
  const config = readModelConfig();
  if (!modelManifest.some((model) => model.id === config.activeModelId)) {
    config.activeModelId = modelManifest[0].id;
  }
  return {
    active_model_id: config.activeModelId,
    cache_dir: modelCacheDir,
    models: modelManifest.map((model) => modelSnapshot(model, config)),
  };
}

function bridgeEnv() {
  const config = readModelConfig();
  const active = modelSnapshot(modelById(config.activeModelId), config);
  return {
    ...process.env,
    MONSTER_GUARD_HOME: guardHome,
    MOONSHINE_MODEL_ID: active.id,
    MOONSHINE_LLM_KNOT_URL: active.url,
    MOONSHINE_SKYMESH_STATUS: skymeshStatusPath,
    FAT_STATION_BIN: join(sidecarDir, 'fat-station'),
    ...(active.present ? { MOONSHINE_LLM_KNOT: active.cache_path } : {}),
    MOONSHINE_EXTRA_PATH: moonshineExtraPath,
    PATH: moonshineExtraPath ? `${moonshineExtraPath}:${process.env.PATH || ''}` : process.env.PATH,
  };
}

function skymeshStatus() {
  const config = readSkymeshConfig();
  const activity = readJson(skymeshStatusPath, {});
  return {
    enabled: process.env.MOONSHINE_SKYMESH_DISABLED !== '1' && config.enabled,
    running: !!skymeshBridge && skymeshBridge.exitCode === null,
    pid: skymeshBridge?.pid ?? null,
    node: typeof activity.node === 'string' ? activity.node : null,
    model: typeof activity.model === 'string' ? activity.model : null,
    last_cache_key: typeof activity.last_cache_key === 'string' ? activity.last_cache_key : typeof activity.lastCacheKey === 'string' ? activity.lastCacheKey : null,
    filled_cache_keys: Array.isArray(activity.filled_cache_keys)
      ? activity.filled_cache_keys.filter((key) => typeof key === 'string')
      : Array.isArray(activity.filledCacheKeys)
        ? activity.filledCacheKeys.filter((key) => typeof key === 'string')
        : [],
    last_error: typeof activity.last_error === 'string' ? activity.last_error : typeof activity.lastError === 'string' ? activity.lastError : null,
    updated_at: typeof activity.updated_at === 'string' ? activity.updated_at : typeof activity.updatedAt === 'string' ? activity.updatedAt : null,
  };
}

function startSkymeshBridge() {
  if (process.env.MOONSHINE_SKYMESH_DISABLED === '1' || !readSkymeshConfig().enabled) return;
  if (skymeshBridge && skymeshBridge.exitCode === null) return;
  if (!statSync(skymeshBridgeScript, { throwIfNoEntry: false })?.isFile()) return;

  skymeshBridge = spawn(process.execPath, [skymeshBridgeScript], {
    env: bridgeEnv(),
    stdio: ['ignore', 'inherit', 'inherit'],
  });
  skymeshBridge.on('exit', (code, signal) => {
    const expected = code === 0 || signal === 'SIGINT' || signal === 'SIGTERM';
    skymeshBridge = null;
    if (!expected && process.env.MOONSHINE_SKYMESH_DISABLED !== '1') {
      setTimeout(startSkymeshBridge, 3000);
    }
  });
}

function restartSkymeshBridge() {
  if (skymeshBridge && skymeshBridge.exitCode === null) {
    skymeshBridge.kill('SIGTERM');
  }
  skymeshBridge = null;
  startSkymeshBridge();
}

function configureSkymesh(enabled) {
  const config = { enabled: enabled !== false };
  writeSkymeshConfig(config);
  if (config.enabled) {
    restartSkymeshBridge();
  } else if (skymeshBridge && skymeshBridge.exitCode === null) {
    skymeshBridge.kill('SIGTERM');
    skymeshBridge = null;
  }
  return skymeshStatus();
}

function configureModel(modelId, url) {
  const model = modelById(modelId);
  const config = readModelConfig();
  config.activeModelId = model.id;
  config.urlOverrides = config.urlOverrides || {};
  if (typeof url === 'string') {
    const trimmed = url.trim();
    if (trimmed) config.urlOverrides[model.id] = trimmed;
    else delete config.urlOverrides[model.id];
  }
  writeModelConfig(config);
  restartSkymeshBridge();
  return modelSnapshot(model, config);
}

async function streamToFile(response, path, append) {
  if (!response.body) throw new Error('download response did not include a body');
  const file = createWriteStream(path, { flags: append ? 'a' : 'w' });
  try {
    for await (const chunk of response.body) {
      if (!file.write(chunk)) await once(file, 'drain');
    }
    file.end();
    await once(file, 'finish');
  } catch (err) {
    file.destroy();
    throw err;
  }
}

async function downloadModel(modelId) {
  const config = readModelConfig();
  const model = modelById(modelId || config.activeModelId);
  const url = modelUrl(model, config);
  const dest = join(modelCacheDir, model.cacheName);
  const partial = join(modelCacheDir, `${model.cacheName}.partial`);
  mkdirSync(modelCacheDir, { recursive: true });

  const existing = statSync(partial, { throwIfNoEntry: false })?.size || 0;
  const headers = existing > 0 ? { range: `bytes=${existing}-` } : {};
  const response = await fetch(url, { headers });
  if (!response.ok) {
    throw new Error(`download failed: ${response.status} ${response.statusText} from ${url}`);
  }

  const resumed = existing > 0 && response.status === 206;
  if (existing > 0 && !resumed) {
    unlinkSync(partial);
  }
  await streamToFile(response, partial, resumed);
  renameSync(partial, dest);

  const updated = { ...config, activeModelId: model.id };
  writeModelConfig(updated);
  restartSkymeshBridge();
  return modelSnapshot(model, updated);
}

function json(res, status, body) {
  const data = JSON.stringify(body);
  res.writeHead(status, {
    'access-control-allow-origin': '*',
    'access-control-allow-methods': 'GET,POST,OPTIONS',
    'access-control-allow-headers': 'authorization,content-type',
    'content-type': 'application/json',
    'content-length': Buffer.byteLength(data),
  });
  res.end(data);
}

function textSnapshot(proc) {
  return {
    id: proc.id,
    command: proc.command,
    status: proc.status,
    created_at: proc.created_at,
    exit_code: proc.exit_code,
    output: proc.output,
    duration_ms: proc.duration_ms,
  };
}

function runProcess(command) {
  const id = randomUUID();
  const startedAt = Date.now();
  const sessionName = `moonshine-mobile-${id}`;
  const proc = {
    id,
    command,
    status: 'running',
    created_at: new Date(startedAt).toISOString(),
    exit_code: null,
    output: '',
    duration_ms: null,
    child: null,
  };
  processes.set(id, proc);

  const delegate = spawn(guardBin, ['delegate', sessionName, 'fs,net,exec', '--ttl', '86400'], {
    env: bridgeEnv(),
  });
  let delegateErr = '';
  delegate.stderr.on('data', (chunk) => {
    delegateErr += chunk.toString();
  });
  delegate.on('close', (code) => {
    if (code !== 0) {
      proc.status = 'completed';
      proc.exit_code = code ?? 1;
      proc.output = `delegate failed: ${delegateErr || `exit ${code}`}`;
      proc.duration_ms = Date.now() - startedAt;
      return;
    }

    const child = spawn(
      guardBin,
      [
        'run',
        '--name',
        sessionName,
        '--caps',
        'fs,net,exec',
        '--lease',
        '86400',
        '--',
        moonshineBin,
        '-c',
        command,
      ],
      { env: bridgeEnv() }
    );
    proc.child = child;
    child.stdout.on('data', (chunk) => {
      proc.output += chunk.toString();
    });
    child.stderr.on('data', (chunk) => {
      proc.output += chunk.toString();
    });
    child.on('close', (code) => {
      proc.status = 'completed';
      proc.exit_code = code ?? 0;
      proc.duration_ms = Date.now() - startedAt;
    });
    child.on('error', (err) => {
      proc.status = 'completed';
      proc.exit_code = 1;
      proc.output += `spawn failed: ${err.message}`;
      proc.duration_ms = Date.now() - startedAt;
    });
  });

  return textSnapshot(proc);
}

async function bodyJson(req) {
  const chunks = [];
  for await (const chunk of req) chunks.push(chunk);
  if (chunks.length === 0) return {};
  return JSON.parse(Buffer.concat(chunks).toString('utf8'));
}

function authorized(req) {
  return req.headers.authorization === `Bearer ${token}`;
}

const server = createServer(async (req, res) => {
  try {
    if (req.method === 'OPTIONS') return json(res, 200, {});
    if (!authorized(req)) return json(res, 401, { error: 'unauthorized' });

    const url = new URL(req.url || '/', `http://${req.headers.host || 'localhost'}`);
    if (req.method === 'GET' && url.pathname === '/health') {
      return json(res, 200, { ok: true, guardBin, moonshineBin, guardHome, model: modelStatus(), skymesh: skymeshStatus() });
    }
    if (req.method === 'GET' && url.pathname === '/models') {
      return json(res, 200, modelStatus());
    }
    if (req.method === 'GET' && url.pathname === '/skymesh') {
      return json(res, 200, skymeshStatus());
    }
    if (req.method === 'POST' && url.pathname === '/configure_skymesh') {
      const body = await bodyJson(req);
      return json(res, 200, configureSkymesh(body.enabled));
    }
    if (req.method === 'POST' && url.pathname === '/configure_model') {
      const body = await bodyJson(req);
      if (typeof body.model_id !== 'string' || body.model_id.trim() === '') {
        return json(res, 400, { error: 'model_id required' });
      }
      return json(res, 200, configureModel(body.model_id, body.url));
    }
    if (req.method === 'POST' && url.pathname === '/download_model') {
      const body = await bodyJson(req);
      return json(res, 200, await downloadModel(body.model_id));
    }
    if (req.method === 'GET' && url.pathname === '/system_info') {
      return json(res, 200, {
        moonshine_bin: moonshineBin,
        guard_bin: guardBin,
        guard_home: guardHome,
        identity_present: true,
        identity_path: `${guardHome} (bridge)`,
        grants_count: 0,
        repl_active: true,
        ready: true,
        problems: [],
      });
    }
    if (req.method === 'GET' && url.pathname === '/processes') {
      return json(
        res,
        200,
        Array.from(processes.values()).map(textSnapshot).reverse()
      );
    }
    if (req.method === 'GET' && url.pathname.startsWith('/process/')) {
      const id = decodeURIComponent(url.pathname.slice('/process/'.length));
      const proc = processes.get(id);
      return proc ? json(res, 200, textSnapshot(proc)) : json(res, 404, { error: 'not found' });
    }
    if (req.method === 'POST' && url.pathname === '/spawn_shell') {
      const body = await bodyJson(req);
      if (typeof body.command !== 'string' || body.command.trim() === '') {
        return json(res, 400, { error: 'command required' });
      }
      return json(res, 200, runProcess(body.command));
    }
    if (req.method === 'POST' && url.pathname === '/clear_history') {
      for (const [id, proc] of processes) {
        if (proc.status !== 'running') processes.delete(id);
      }
      return json(res, 200, {});
    }
    if (req.method === 'POST' && url.pathname === '/interrupt') {
      for (const proc of processes.values()) {
        if (proc.status === 'running' && proc.child) proc.child.kill('SIGINT');
      }
      return json(res, 200, {});
    }
    if (req.method === 'POST' && url.pathname === '/complete') {
      return json(res, 200, []);
    }
    return json(res, 404, { error: 'not found' });
  } catch (err) {
    return json(res, 500, { error: err instanceof Error ? err.message : String(err) });
  }
});

function isMainModule() {
  try {
    return !!process.argv[1] && realpathSync(process.argv[1]) === realpathSync(fileURLToPath(import.meta.url));
  } catch {
    return false;
  }
}

function startMobileBridge() {
  server.listen(port, host, () => {
    console.log(`moonshine mobile bridge listening on http://${host}:${port}`);
    console.log(`MOONSHINE_MOBILE_BRIDGE_TOKEN=${token}`);
    console.log(`moonshine=${moonshineBin}`);
    console.log(`monster-guard=${guardBin}`);
    console.log(`MOONSHINE_EXTRA_PATH=${moonshineExtraPath}`);
    console.log(`MOONSHINE_MODEL_CACHE_DIR=${modelCacheDir}`);
    startSkymeshBridge();
  });

  process.on('SIGINT', () => {
    skymeshBridge?.kill('SIGINT');
    process.exit(0);
  });
  process.on('SIGTERM', () => {
    skymeshBridge?.kill('SIGTERM');
    process.exit(0);
  });
}

if (isMainModule()) {
  startMobileBridge();
}

export {
  authorized,
  bridgeEnv,
  configureModel,
  configureSkymesh,
  modelById,
  modelStatus,
  readJson,
  readSkymeshConfig,
  skymeshStatus,
  startMobileBridge,
};
