// @a0n/gnosis/src/gnosis-time/clock.ts
/**
 * GnosisTime clock — discrete **`Fin 12`** dial (**`+1 (mod 12)`**) aligned with Lean **`Gnosis.GnosisTimeClock`**
 * (`open-source/gnosis-math/Gnosis/GnosisTimeClock.lean`) and **`Gnosis.Circadian`** constants.
 *
 * Carrier only: nonnegative integer counters (Nat-shaped); no continuum, no biological drift semantics here.
 *
 * **φ rotation:** **`tickPhiGated`** advances an internal **`θ ∈ [0,1)`** by **`φ − 1 = φ⁻¹`** when the gate opens — the classical
 * irrational rotation / Kronecker step on **`ℝ/ℤ`**. The twelve-phase readout **`⌊12 θ⌋`** coarsens that flow to **`Fin 12`**.
 * Floating-point **`θ`** is a surrogate only (**exact** arithmetic would need another carrier).
 */

/** Matches **`Gnosis.Circadian.aeon`** (`12`). */
export const AEON = 12;

/** Alias matching **`AeonCycleTwelveShadow.twelve`**. */
export const TWELVE = AEON;

/** Matches **`Gnosis.Circadian.primitives`** (`5`). */
export const PRIMITIVES = 5;

/** Matches **`Gnosis.Circadian.syzygy`** (`2`). */
export const SYZYGY = 2;

/** Matches **`Gnosis.Circadian.kenoma`** (`10`). */
export const KENOMA = 10;

/** `primitives * aeon` — matches **`Circadian.minutesPerHour`**. */
export const MINUTES_PER_HOUR = PRIMITIVES * AEON;

/** Golden ratio **`φ = (1 + √5) / 2`**. */
export const PHI = (1 + Math.sqrt(5)) / 2;

/**
 * Irrational rotation increment **`α = φ − 1 = φ⁻¹`** on **`[0,1)`** with **`+ mod 1`**.
 * Satisfies **`φ · α = 1`** exactly in **`ℝ`** (surrogate: IEEE double).
 */
export const PHI_APERIODIC_ROTATION_STEP = PHI - 1;

/** Internal angle for **`tickPhi` / **`tickPhiGated`** (fractional surrogate on **`ℝ/ℤ`**). */
export type PhiRotationState = Readonly<{
  theta: number;
}>;

/** Phase index **`0 … aeon - 1`** (twelve-hour discrete dial). */
export type TimePhase = number;

/** Whether **`n`** is a valid **`TimePhase`**. */
export function isValidTimePhase(n: number): boolean {
  return Number.isInteger(n) && n >= 0 && n < AEON;
}

/** Require a **`TimePhase`** or throw (**fail-closed API surface**). */
export function timePhase(n: number): TimePhase {
  if (!isValidTimePhase(n)) {
    throw new RangeError(
      `GnosisTime timePhase expects integer in [0, ${AEON}), got ${String(n)}`
    );
  }
  return n;
}

/** Map **`x`** to **`[0, 1)`** (**`mod 1`**). */
export function normalizeUnitFrac(x: number): number {
  if (!Number.isFinite(x)) {
    throw new RangeError(
      `GnosisTime normalizeUnitFrac expects finite x, got ${String(x)}`
    );
  }
  let y = x % 1;
  if (y < 0) {
    y += 1;
  }
  return y;
}

/** Canonical **`PhiRotationState`** from any finite angle. */
export function phiRotationState(theta: number): PhiRotationState {
  return { theta: normalizeUnitFrac(theta) };
}

/** Twelve-slot phase readout **`⌊ aeon · θ ⌋`** clamped to **`Fin aeon`**. */
export function phaseFromPhiTheta(theta: number): TimePhase {
  const t = normalizeUnitFrac(theta);
  const slot = Math.min(AEON - 1, Math.max(0, Math.floor(t * AEON)));
  return timePhase(slot);
}

/**
 * **Gated φ tick:** when **`gateOpen`**, advance **`θ`** by **`PHI_APERIODIC_ROTATION_STEP`** (**`mod 1`**); when closed, **`θ`** is frozen.
 * Always returns the coarse **`TimePhase`** from the current **`θ`** after the optional advance.
 */
export function tickPhiGated(
  state: PhiRotationState,
  gateOpen: boolean
): { state: PhiRotationState; phase: TimePhase } {
  const theta0 = normalizeUnitFrac(state.theta);
  if (!gateOpen) {
    return { state: { theta: theta0 }, phase: phaseFromPhiTheta(theta0) };
  }
  const theta1 = normalizeUnitFrac(theta0 + PHI_APERIODIC_ROTATION_STEP);
  return { state: { theta: theta1 }, phase: phaseFromPhiTheta(theta1) };
}

/** Ungated **`tickPhiGated(state, true)`** — one **`φ`** rotation step. */
export function tickPhi(state: PhiRotationState): {
  state: PhiRotationState;
  phase: TimePhase;
} {
  return tickPhiGated(state, true);
}

/** One **`cyclicSucc`** tick (**matches **`GnosisTimeClock.tick`**). */
export function tick(t: TimePhase): TimePhase {
  timePhase(t);
  return (t + 1) % AEON;
}

/** **`k`** composed ticks (**matches **`tickIterate` val-level **`(t.val + k) % twelve`**). */
export function tickIterate(k: number, t: TimePhase): TimePhase {
  timePhase(t);
  if (!Number.isInteger(k) || k < 0) {
    throw new RangeError(
      `GnosisTime tickIterate expects nonnegative integer k, got ${String(k)}`
    );
  }
  return (t + k) % AEON;
}

/** **`aeon`** ticks restore the phase (**global period**). */
export function tickIterateAeon(t: TimePhase): TimePhase {
  return tickIterate(AEON, t);
}

/** **`m % aeon`** for nonnegative **`m`** (**`phaseOfNatTick`**). */
export function phaseOfNatTick(m: number): TimePhase {
  if (!Number.isInteger(m) || m < 0) {
    throw new RangeError(
      `GnosisTime phaseOfNatTick expects nonnegative integer m, got ${String(m)}`
    );
  }
  return m % AEON;
}

/** Adding **`k · aeon`** does not move **`phaseOfNatTick`** (**`ℤ/12ℤ`** stability). */
export function phaseOfNatTickAddAeonMul(m: number, k: number): TimePhase {
  if (!Number.isInteger(m) || m < 0) {
    throw new RangeError(
      `GnosisTime phaseOfNatTickAddAeonMul expects nonnegative integer m, got ${String(m)}`
    );
  }
  if (!Number.isInteger(k) || k < 0) {
    throw new RangeError(
      `GnosisTime phaseOfNatTickAddAeonMul expects nonnegative integer k, got ${String(k)}`
    );
  }
  return phaseOfNatTick(m + k * AEON);
}

/** Stacked solar hours leave **`phaseOfNatTick`** fixed (**`minutesPerHour`** multiple of **`aeon`**). */
export function phaseOfNatTickAddMinutesPerHourMul(m: number, h: number): TimePhase {
  if (!Number.isInteger(m) || m < 0) {
    throw new RangeError(
      `GnosisTime phaseOfNatTickAddMinutesPerHourMul expects nonnegative integer m, got ${String(m)}`
    );
  }
  if (!Number.isInteger(h) || h < 0) {
    throw new RangeError(
      `GnosisTime phaseOfNatTickAddMinutesPerHourMul expects nonnegative integer h, got ${String(h)}`
    );
  }
  return phaseOfNatTick(m + h * MINUTES_PER_HOUR);
}

/** **`minutesPerHour ≡ 0 (mod aeon)`** certificate. */
export function minutesPerHourModAeonZero(): boolean {
  return MINUTES_PER_HOUR % AEON === 0;
}

/** Identity bridge — same twelve-slot index as **`TwelveSlot`** / **`pairsIJ`** vertices in Lean. */
export function twelveSlotOfPhase(t: TimePhase): TimePhase {
  return timePhase(t);
}
