// @a0n/gnosis/src/gnosis-time/triton-admission-gate.ts
import { Verdict } from '@a0n/bitwise/triton-verdict';

import { AEON, phaseOfNatTick, type TimePhase, timePhase } from './clock';

export const AEON_PAIR_CARRIER_CAPACITY = (AEON * (AEON - 1)) / 2;
export const BWTR_MAGIC = 'BWTR';
export const BWTR_MAGIC_BYTES = new Uint8Array([0x42, 0x57, 0x54, 0x52]);
export const BWTR_VERSION = 1;
export const BWTR_ABSENT_U8 = 0xff;
export const BWTR_ABSENT_U64 = 0xffff_ffff_ffff_ffffn;
export const BWTR_HEADER_LEN = 18;

export type TritonAdmissionPhase =
  | 'fork-future'
  | 'race-fold-present'
  | 'vent-interfere-past';

export type TritonAdmissionStatus = 'admit' | 'defer' | 'reject';

/**
 * The back-compat string status, re-expressed via the canonical {@link Verdict}.
 * This site was already triton-shaped (admit/defer/reject); unifying it proves
 * the canonical type drops in without changing behavior:
 *   - admit  ≅ Accept   (+1)
 *   - defer  ≅ Abstain   (0, the neutral middle)
 *   - reject ≅ Decline  (-1)
 */
export const TRITON_ADMISSION_VERDICT: Readonly<
  Record<TritonAdmissionStatus, Verdict>
> = {
  admit: Verdict.Accept,
  defer: Verdict.Abstain,
  reject: Verdict.Decline,
};

/** The canonical verdict for a string status (admit/defer/reject → Verdict). */
export function verdictForTritonAdmissionStatus(
  status: TritonAdmissionStatus
): Verdict {
  return TRITON_ADMISSION_VERDICT[status];
}

/** The back-compat string status for a canonical verdict (inverse bridge). */
export function tritonAdmissionStatusForVerdict(
  verdict: Verdict
): TritonAdmissionStatus {
  switch (verdict) {
    case Verdict.Accept:
      return 'admit';
    case Verdict.Abstain:
      return 'defer';
    case Verdict.Decline:
      return 'reject';
  }
}

export type TritonAdmissionWork = Readonly<{
  phase: TritonAdmissionPhase;
  clockPhase: TimePhase;
  carrierKey?: number;
  forkBudget?: number;
  seenCanonicalC2Keys?: ReadonlySet<number>;
}>;

export type TritonAdmissionResult = Readonly<{
  status: TritonAdmissionStatus;
  /**
   * The canonical verdict — the single shared grid type. Derived from `status`
   * (admit=Accept, defer=Abstain, reject=Decline) so the grid speaks one
   * verdict alphabet while the string status stays for back-compat.
   */
  verdict: Verdict;
  reason: string;
  clockPhase: TimePhase;
  expectedClockPhase: TimePhase;
  carrierKey?: number;
  canonicalC2Key?: number;
}>;

export type ResidentTritonPayload = Readonly<{
  work: TritonAdmissionWork;
  userId: string;
}>;

export type BwtrFixture = Readonly<{
  name: string;
  userId: string;
  work: TritonAdmissionWork;
  hex: string;
}>;

export const TRITON_PHASE_CLOCK_SLOT: Readonly<Record<TritonAdmissionPhase, TimePhase>> =
  {
    'race-fold-present': 0,
    'fork-future': 1,
    'vent-interfere-past': AEON - 1,
  };

/**
 * Handles the gnosis clock Phase For Triton Admission workflow.
 */
export function clockPhaseForTritonAdmission(
  phase: TritonAdmissionPhase
): TimePhase {
  return TRITON_PHASE_CLOCK_SLOT[phase];
}

/**
 * Handles the gnosis aeon Pair Carrier Key workflow.
 */
export function aeonPairCarrierKey(left: number, right: number): number {
  const i = timePhase(left);
  const j = timePhase(right);
  if (i >= j) {
    throw new RangeError(
      `aeonPairCarrierKey expects strict pair 0 <= left < right < ${AEON}, got ${left}, ${right}`
    );
  }
  let rowOffset = 0;
  for (let row = 0; row < i; row++) {
    rowOffset += AEON - row - 1;
  }
  return rowOffset + (j - i - 1);
}

/**
 * Returns whether is Valid Aeon Pair Carrier Key is true.
 */
export function isValidAeonPairCarrierKey(key: number): boolean {
  return (
    Number.isInteger(key) &&
    key >= 0 &&
    key < AEON_PAIR_CARRIER_CAPACITY
  );
}

/**
 * Handles the gnosis aeon Pair Carrier Key Or Throw workflow.
 */
export function aeonPairCarrierKeyOrThrow(key: number): number {
  if (!isValidAeonPairCarrierKey(key)) {
    throw new RangeError(
      `Aeon carrier key expects integer in [0, ${AEON_PAIR_CARRIER_CAPACITY}), got ${String(key)}`
    );
  }
  return key;
}

/**
 * Handles the gnosis c2 Swap Boundary Carrier Key workflow.
 */
export function c2SwapBoundaryCarrierKey(key: number): number {
  const carrierKey = aeonPairCarrierKeyOrThrow(key);
  if (carrierKey === 0) {
    return 1;
  }
  if (carrierKey === 1) {
    return 0;
  }
  return carrierKey;
}

/**
 * Handles the gnosis canonical C2 Carrier Key workflow.
 */
export function canonicalC2CarrierKey(key: number): number {
  const carrierKey = aeonPairCarrierKeyOrThrow(key);
  return Math.min(carrierKey, c2SwapBoundaryCarrierKey(carrierKey));
}

/**
 * Handles the gnosis evaluate Triton Admission workflow.
 */
export function evaluateTritonAdmission(
  work: TritonAdmissionWork
): TritonAdmissionResult {
  const expectedClockPhase = clockPhaseForTritonAdmission(work.phase);
  const clockPhase = timePhase(work.clockPhase);

  // Single result constructor: derives the canonical `verdict` from `status`
  // so admit/defer/reject and Accept/Abstain/Decline can never drift apart.
  const result = (
    status: TritonAdmissionStatus,
    reason: string,
    extra?: { carrierKey?: number; canonicalC2Key?: number }
  ): TritonAdmissionResult => ({
    status,
    verdict: verdictForTritonAdmissionStatus(status),
    reason,
    clockPhase,
    expectedClockPhase,
    ...(extra?.carrierKey === undefined ? {} : { carrierKey: extra.carrierKey }),
    ...(extra?.canonicalC2Key === undefined
      ? {}
      : { canonicalC2Key: extra.canonicalC2Key }),
  });

  if (clockPhase !== expectedClockPhase) {
    return result('reject', 'triton-clock-phase-mismatch');
  }

  if (work.phase === 'fork-future' && work.forkBudget !== undefined) {
    if (!Number.isFinite(work.forkBudget)) {
      return result('reject', 'fork-budget-non-finite');
    }
    if (work.forkBudget <= 0) {
      return result('defer', 'fork-budget-exhausted');
    }
  }

  if (work.carrierKey === undefined) {
    return result('admit', 'triton-clock-admitted');
  }

  if (!isValidAeonPairCarrierKey(work.carrierKey)) {
    return result('reject', 'aeon-carrier-key-out-of-range');
  }

  const carrierKey = work.carrierKey;
  const canonicalC2Key = canonicalC2CarrierKey(carrierKey);
  if (work.seenCanonicalC2Keys?.has(canonicalC2Key) === true) {
    return result('reject', 'c2-duplicate-carrier', { carrierKey, canonicalC2Key });
  }

  return result('admit', 'triton-carrier-admitted', { carrierKey, canonicalC2Key });
}

/**
 * Handles the gnosis resident Gnosis Clock Phase Of Nat Tick workflow.
 */
export function residentGnosisClockPhaseOfNatTick(sequence: number): TimePhase {
  return phaseOfNatTick(sequence);
}

/**
 * Handles the gnosis encode Resident Binary Triton Payload workflow.
 */
export function encodeResidentBinaryTritonPayload(
  work: TritonAdmissionWork,
  userId: string
): Uint8Array {
  const userBytes = new TextEncoder().encode(userId);
  const payload = new Uint8Array(BWTR_HEADER_LEN + userBytes.length);
  payload.set(BWTR_MAGIC_BYTES, 0);
  payload[4] = BWTR_VERSION;
  payload[5] = tritonPhaseToWire(work.phase);
  payload[6] = timePhase(work.clockPhase);
  payload[7] =
    work.carrierKey === undefined ? BWTR_ABSENT_U8 : u8OrThrow(work.carrierKey);
  payload[8] =
    work.seenCanonicalC2Keys === undefined
      ? BWTR_ABSENT_U8
      : u8OrThrow(firstSeenCanonicalC2Key(work.seenCanonicalC2Keys));
  payload[9] = 0;
  const forkBudget =
    work.forkBudget === undefined ? BWTR_ABSENT_U64 : u64OrThrow(work.forkBudget);
  new DataView(payload.buffer, payload.byteOffset, payload.byteLength).setBigUint64(
    10,
    forkBudget,
    false
  );
  payload.set(userBytes, BWTR_HEADER_LEN);
  return payload;
}

/**
 * Parses the Resident Triton Payload.
 */
export function parseResidentTritonPayload(
  payload: Uint8Array,
  sequence: number
): ResidentTritonPayload {
  if (hasBwtrMagic(payload)) {
    return parseResidentBinaryTritonPayload(payload);
  }

  const text = new TextDecoder('utf-8', { fatal: true }).decode(payload);
  if (!text.startsWith('triton:')) {
    return {
      work: {
        phase: 'race-fold-present',
        clockPhase: residentGnosisClockPhaseOfNatTick(sequence),
      },
      userId: text,
    };
  }

  const separator = text.indexOf(';');
  if (separator < 0) {
    throw new Error("resident Triton payload missing ';' separator");
  }

  const header = text.slice('triton:'.length, separator);
  const userId = text.slice(separator + 1);
  let phase: TritonAdmissionPhase = 'race-fold-present';
  let clockPhase: TimePhase | undefined;
  let carrierKey: number | undefined;
  let forkBudget: number | undefined;
  let seenCanonicalC2Key: number | undefined;

  for (const rawEntry of header.split(',')) {
    const entry = rawEntry.trim();
    if (entry.length === 0) {
      continue;
    }
    const [key, rawValue, extra] = entry.split('=');
    if (key === undefined || rawValue === undefined || extra !== undefined) {
      throw new Error(`resident Triton metadata entry missing '=': ${entry}`);
    }
    const value = rawValue.trim();
    switch (key.trim()) {
      case 'phase':
        phase = parseTritonPhase(value);
        break;
      case 'clock':
      case 'clockPhase':
        clockPhase = timePhase(parseFiniteInteger(value, 'Triton clock phase'));
        break;
      case 'carrier':
      case 'carrierKey':
        carrierKey = parseFiniteInteger(value, 'Triton carrier key');
        break;
      case 'budget':
      case 'forkBudget':
        forkBudget = parseFiniteInteger(value, 'Triton fork budget');
        break;
      case 'seen':
      case 'seenC2':
      case 'seenCanonicalC2Key':
        seenCanonicalC2Key = parseFiniteInteger(value, 'Triton seen C2 key');
        break;
      default:
        throw new Error(`unknown resident Triton metadata key '${key.trim()}'`);
    }
  }

  return {
    work: {
      phase,
      clockPhase: clockPhase ?? residentGnosisClockPhaseOfNatTick(sequence),
      ...(carrierKey === undefined ? {} : { carrierKey }),
      ...(forkBudget === undefined ? {} : { forkBudget }),
      ...(seenCanonicalC2Key === undefined
        ? {}
        : { seenCanonicalC2Keys: new Set([seenCanonicalC2Key]) }),
    },
    userId,
  };
}

export const BWTR_FIXTURES: readonly BwtrFixture[] = [
  {
    name: 'fork-future-carrier-user-789',
    userId: 'user-789',
    work: {
      phase: 'fork-future',
      clockPhase: 1,
      carrierKey: 11,
      forkBudget: 2,
      seenCanonicalC2Keys: new Set([3]),
    },
    hex: '425754520101010b03000000000000000002757365722d373839',
  },
];

function parseResidentBinaryTritonPayload(
  payload: Uint8Array
): ResidentTritonPayload {
  if (payload.length < BWTR_HEADER_LEN) {
    throw new Error('resident binary Triton payload shorter than fixed header');
  }
  if (payload[4] !== BWTR_VERSION) {
    throw new Error(`unsupported resident binary Triton version ${payload[4]}`);
  }

  const view = new DataView(payload.buffer, payload.byteOffset, payload.byteLength);
  const rawBudget = view.getBigUint64(10, false);
  const carrierKey = payload[7] === BWTR_ABSENT_U8 ? undefined : payload[7];
  const seenC2Key = payload[8] === BWTR_ABSENT_U8 ? undefined : payload[8];
  const userId = new TextDecoder('utf-8', { fatal: true }).decode(
    payload.slice(BWTR_HEADER_LEN)
  );

  return {
    work: {
      phase: tritonPhaseFromWire(payload[5] ?? BWTR_ABSENT_U8),
      clockPhase: timePhase(payload[6] ?? BWTR_ABSENT_U8),
      ...(carrierKey === undefined ? {} : { carrierKey }),
      ...(rawBudget === BWTR_ABSENT_U64 ? {} : { forkBudget: Number(rawBudget) }),
      ...(seenC2Key === undefined ? {} : { seenCanonicalC2Keys: new Set([seenC2Key]) }),
    },
    userId,
  };
}

function hasBwtrMagic(payload: Uint8Array): boolean {
  return BWTR_MAGIC_BYTES.every((byte, index) => payload[index] === byte);
}

function tritonPhaseToWire(phase: TritonAdmissionPhase): number {
  switch (phase) {
    case 'race-fold-present':
      return 0;
    case 'fork-future':
      return 1;
    case 'vent-interfere-past':
      return 2;
  }
}

function tritonPhaseFromWire(value: number): TritonAdmissionPhase {
  switch (value) {
    case 0:
      return 'race-fold-present';
    case 1:
      return 'fork-future';
    case 2:
      return 'vent-interfere-past';
    default:
      throw new Error(`unknown binary Triton phase tag ${value}`);
  }
}

function parseTritonPhase(value: string): TritonAdmissionPhase {
  switch (value) {
    case 'fork-future':
    case 'future':
    case 'fork':
      return 'fork-future';
    case 'race-fold-present':
    case 'present':
    case 'race-fold':
      return 'race-fold-present';
    case 'vent-interfere-past':
    case 'past':
    case 'vent':
      return 'vent-interfere-past';
    default:
      throw new Error(`unknown Triton phase '${value}'`);
  }
}

function firstSeenCanonicalC2Key(keys: ReadonlySet<number>): number {
  const [first] = keys;
  if (first === undefined) {
    return BWTR_ABSENT_U8;
  }
  return first;
}

function u8OrThrow(value: number): number {
  if (!Number.isInteger(value) || value < 0 || value > 0xff) {
    throw new RangeError(`BWTR u8 field expects integer in [0, 255], got ${String(value)}`);
  }
  return value;
}

function u64OrThrow(value: number): bigint {
  if (!Number.isSafeInteger(value) || value < 0) {
    throw new RangeError(`BWTR u64 field expects safe nonnegative integer, got ${String(value)}`);
  }
  return BigInt(value);
}

function parseFiniteInteger(value: string, field: string): number {
  const parsed = Number(value);
  if (!Number.isInteger(parsed)) {
    throw new Error(`invalid ${field} '${value}'`);
  }
  return parsed;
}
