// @a0n/.ignored_gnosis/src/benchmarks/aeon-unspeech-hybrid-benchmark.ts
#!/usr/bin/env bun
/**
 * Aeon Unspeech Hybrid Benchmark
 *
 * Evaluates lossless state transport over Aeon-style framing (10-byte header
 * per frame, out-of-order reassembly across channels) for:
 * - Brotli full-state per tick
 * - Unspeech hybrid (Brotli keyframes + UTF-16 teleport deltas)
 *
 * The benchmark explicitly shreds each message into frame payload chunks,
 * assigns chunks across multiple channels, shuffles arrival order, reassembles,
 * and verifies exact state recovery.
 */

import {
  brotliCompressSync,
  brotliDecompressSync,
  constants as zlibConstants,
} from 'node:zlib';
import { pathToFileURL } from 'node:url';
import {
  buildUtf16Lexicon,
  teleportUtf16ThroughLexicon,
  inflateUtf16FromTeleport,
  type UnicodeTeleportPacket,
  type UnicodeTeleportMode,
  type UnicodeReorderMode,
} from '../unspeech-unicode-teleport';
import type { TritonSelfSimilarPacket } from '../unspeech-triton';

const AEON_HEADER_BYTES = 10;
const KEYFRAME_MESSAGE_TAG = 0;
const TELEPORT_MESSAGE_TAG = 1;

const TELEPORT_DIRECT_TAG = 1;
const TELEPORT_DIRECT_REORDERED_TAG = 2;
const TELEPORT_RESIDUAL_TAG = 3;

const SIGN_ENCODING_RLE_TAG = 0;
const SIGN_ENCODING_FOLDED_TAG = 1;

interface TransportConfig {
  channels: number;
  framePayloadBytes: number;
}

interface AeonFrameChunk {
  channel: number;
  sequence: number;
  payload: Uint8Array;
}

interface Scenario {
  name: string;
  states: string[];
}

interface StrategyConfig {
  name: string;
  kind: 'brotli-full' | 'brotli-full-utf8' | 'unspeech-hybrid';
  keyframeInterval?: number;
  teleportMode?: UnicodeTeleportMode;
  reorderMode?: UnicodeReorderMode;
}

interface StrategyResult {
  strategy: string;
  rawUtf16Bytes: number;
  wireBytes: number;
  frameCount: number;
  avgFramesPerTick: number;
  compressionSavingsPct: number;
  packetKinds: Record<string, number>;
}

interface ScenarioResult {
  scenario: string;
  tickCount: number;
  results: StrategyResult[];
  winner: StrategyResult;
}

interface VarUintRead {
  value: number;
  nextOffset: number;
}

interface DecodedMessage {
  kind: 'keyframe' | 'teleport';
  keyframePayload?: Uint8Array;
  teleportPacket?: UnicodeTeleportPacket;
}

function createLcg(seed: number): () => number {
  let state = seed >>> 0;
  return () => {
    state = (Math.imul(state, 1664525) + 1013904223) >>> 0;
    return state / 0x100000000;
  };
}

function writeVarUint(output: number[], value: number): void {
  if (!Number.isInteger(value) || value < 0) {
    throw new RangeError(`varuint expects non-negative integer, received ${value}`);
  }
  let remainder = value;
  do {
    let byte = remainder % 128;
    remainder = Math.floor(remainder / 128);
    if (remainder > 0) byte += 128;
    output.push(byte);
  } while (remainder > 0);
}

function readVarUint(input: Uint8Array, startOffset: number): VarUintRead {
  let value = 0;
  let factor = 1;
  let offset = startOffset;
  for (let i = 0; i < 8; i++) {
    if (offset >= input.length) {
      throw new RangeError('unexpected end of payload while reading varuint');
    }
    const byte = input[offset]!;
    offset += 1;
    value += (byte & 0x7f) * factor;
    if ((byte & 0x80) === 0) {
      return { value, nextOffset: offset };
    }
    factor *= 128;
  }
  throw new RangeError('varuint exceeds supported length');
}

function concatUint8Arrays(chunks: readonly Uint8Array[]): Uint8Array {
  const total = chunks.reduce((sum, chunk) => sum + chunk.length, 0);
  const out = new Uint8Array(total);
  let offset = 0;
  for (const chunk of chunks) {
    out.set(chunk, offset);
    offset += chunk.length;
  }
  return out;
}

function utf16leBytes(text: string): Uint8Array {
  return new Uint8Array(Buffer.from(text, 'utf16le'));
}

function fromUtf16leBytes(bytes: Uint8Array): string {
  return Buffer.from(bytes).toString('utf16le');
}

function utf8Bytes(text: string): Uint8Array {
  return new Uint8Array(Buffer.from(text, 'utf8'));
}

function fromUtf8Bytes(bytes: Uint8Array): string {
  return Buffer.from(bytes).toString('utf8');
}

function encodeSignPacket(
  out: number[],
  packet: TritonSelfSimilarPacket
): void {
  out.push(
    packet.encoding === 'self-similar-rle'
      ? SIGN_ENCODING_RLE_TAG
      : SIGN_ENCODING_FOLDED_TAG
  );
  writeVarUint(out, packet.coreLength);
  writeVarUint(out, packet.tritonLength);
  writeVarUint(out, packet.payload.length);
  for (const byte of packet.payload) out.push(byte);
}

function decodeSignPacket(
  payload: Uint8Array,
  offset: number
): { packet: TritonSelfSimilarPacket; nextOffset: number } {
  if (offset >= payload.length) {
    throw new RangeError('unexpected end of payload while reading sign encoding');
  }
  const encodingTag = payload[offset]!;
  offset += 1;
  const coreLengthField = readVarUint(payload, offset);
  offset = coreLengthField.nextOffset;
  const tritonLengthField = readVarUint(payload, offset);
  offset = tritonLengthField.nextOffset;
  const signLenField = readVarUint(payload, offset);
  offset = signLenField.nextOffset;
  const signEnd = offset + signLenField.value;
  if (signEnd > payload.length) {
    throw new RangeError('sign packet payload exceeds message bounds');
  }
  const signPayload = payload.subarray(offset, signEnd);
  offset = signEnd;

  const motifSize = signPayload.length >= 6 ? signPayload[5]! : 5;
  const encoding =
    encodingTag === SIGN_ENCODING_FOLDED_TAG
      ? 'self-similar-rle+bitwise-fold'
      : 'self-similar-rle';

  return {
    packet: {
      encoding,
      motifSize,
      tritonLength: tritonLengthField.value,
      rawLength: tritonLengthField.value,
      coreLength: coreLengthField.value,
      payloadLength: signPayload.length,
      compressionRatio:
        tritonLengthField.value === 0
          ? 1
          : tritonLengthField.value / Math.max(1, signPayload.length),
      coreFingerprintHex: '',
      payloadFingerprintHex: '',
      payload: signPayload,
    },
    nextOffset: offset,
  };
}

function serializeTeleportPacket(packet: UnicodeTeleportPacket): Uint8Array {
  const out: number[] = [];

  if (packet.strategy === 'direct') {
    out.push(TELEPORT_DIRECT_TAG);
    writeVarUint(out, packet.length);
    writeVarUint(out, packet.firstIndex);
    encodeSignPacket(out, packet.signPacket);
    writeVarUint(out, packet.magnitudeBytes.length);
    for (const byte of packet.magnitudeBytes) out.push(byte);
    return new Uint8Array(out);
  }

  if (packet.strategy === 'direct-reordered') {
    out.push(TELEPORT_DIRECT_REORDERED_TAG);
    writeVarUint(out, packet.length);
    writeVarUint(out, packet.firstIndex);
    encodeSignPacket(out, packet.signPacket);
    writeVarUint(out, packet.magnitudeBytes.length);
    for (const byte of packet.magnitudeBytes) out.push(byte);
    writeVarUint(out, packet.permutationBytes.length);
    for (const byte of packet.permutationBytes) out.push(byte);
    return new Uint8Array(out);
  }

  out.push(TELEPORT_RESIDUAL_TAG);
  writeVarUint(out, packet.baselineLength);
  writeVarUint(out, packet.targetLength);
  writeVarUint(out, packet.patchCount);
  writeVarUint(out, packet.patchBytes.length);
  for (const byte of packet.patchBytes) out.push(byte);
  return new Uint8Array(out);
}

function deserializeTeleportPacket(payload: Uint8Array): UnicodeTeleportPacket {
  if (payload.length === 0) {
    throw new RangeError('empty teleport payload');
  }
  let offset = 0;
  const strategyTag = payload[offset]!;
  offset += 1;

  if (strategyTag === TELEPORT_DIRECT_TAG) {
    const lengthField = readVarUint(payload, offset);
    offset = lengthField.nextOffset;
    const firstIndexField = readVarUint(payload, offset);
    offset = firstIndexField.nextOffset;
    const signDecoded = decodeSignPacket(payload, offset);
    offset = signDecoded.nextOffset;
    const magnitudeLenField = readVarUint(payload, offset);
    offset = magnitudeLenField.nextOffset;
    const magnitudeEnd = offset + magnitudeLenField.value;
    if (magnitudeEnd > payload.length) {
      throw new RangeError('direct magnitude bytes exceed payload bounds');
    }
    const magnitudeBytes = payload.subarray(offset, magnitudeEnd);
    offset = magnitudeEnd;
    if (offset !== payload.length) {
      throw new RangeError('unexpected trailing bytes in direct teleport payload');
    }
    return {
      strategy: 'direct',
      length: lengthField.value,
      firstIndex: firstIndexField.value,
      signPacket: signDecoded.packet,
      magnitudeBytes,
      estimatedPayloadBytes: payload.length,
    };
  }

  if (strategyTag === TELEPORT_DIRECT_REORDERED_TAG) {
    const lengthField = readVarUint(payload, offset);
    offset = lengthField.nextOffset;
    const firstIndexField = readVarUint(payload, offset);
    offset = firstIndexField.nextOffset;
    const signDecoded = decodeSignPacket(payload, offset);
    offset = signDecoded.nextOffset;
    const magnitudeLenField = readVarUint(payload, offset);
    offset = magnitudeLenField.nextOffset;
    const magnitudeEnd = offset + magnitudeLenField.value;
    if (magnitudeEnd > payload.length) {
      throw new RangeError('direct-reordered magnitude bytes exceed bounds');
    }
    const magnitudeBytes = payload.subarray(offset, magnitudeEnd);
    offset = magnitudeEnd;
    const permLenField = readVarUint(payload, offset);
    offset = permLenField.nextOffset;
    const permEnd = offset + permLenField.value;
    if (permEnd > payload.length) {
      throw new RangeError('direct-reordered permutation bytes exceed bounds');
    }
    const permutationBytes = payload.subarray(offset, permEnd);
    offset = permEnd;
    if (offset !== payload.length) {
      throw new RangeError(
        'unexpected trailing bytes in direct-reordered teleport payload'
      );
    }
    return {
      strategy: 'direct-reordered',
      length: lengthField.value,
      firstIndex: firstIndexField.value,
      signPacket: signDecoded.packet,
      magnitudeBytes,
      permutationBytes,
      estimatedPayloadBytes: payload.length,
    };
  }

  if (strategyTag !== TELEPORT_RESIDUAL_TAG) {
    throw new RangeError(`unsupported teleport strategy tag ${strategyTag}`);
  }
  const baselineLenField = readVarUint(payload, offset);
  offset = baselineLenField.nextOffset;
  const targetLenField = readVarUint(payload, offset);
  offset = targetLenField.nextOffset;
  const patchCountField = readVarUint(payload, offset);
  offset = patchCountField.nextOffset;
  const patchLenField = readVarUint(payload, offset);
  offset = patchLenField.nextOffset;
  const patchEnd = offset + patchLenField.value;
  if (patchEnd > payload.length) {
    throw new RangeError('residual patch bytes exceed payload bounds');
  }
  const patchBytes = payload.subarray(offset, patchEnd);
  offset = patchEnd;
  if (offset !== payload.length) {
    throw new RangeError('unexpected trailing bytes in residual teleport payload');
  }
  return {
    strategy: 'residual',
    baselineLength: baselineLenField.value,
    targetLength: targetLenField.value,
    patchCount: patchCountField.value,
    patchBytes,
    estimatedPayloadBytes: payload.length,
  };
}

function serializeKeyframeMessage(compressedUtf16: Uint8Array): Uint8Array {
  const out: number[] = [KEYFRAME_MESSAGE_TAG];
  writeVarUint(out, compressedUtf16.length);
  for (const byte of compressedUtf16) out.push(byte);
  return new Uint8Array(out);
}

function serializeTeleportMessage(packet: UnicodeTeleportPacket): Uint8Array {
  const packetBytes = serializeTeleportPacket(packet);
  const out: number[] = [TELEPORT_MESSAGE_TAG];
  writeVarUint(out, packetBytes.length);
  for (const byte of packetBytes) out.push(byte);
  return new Uint8Array(out);
}

function deserializeMessage(payload: Uint8Array): DecodedMessage {
  if (payload.length === 0) {
    throw new RangeError('empty message payload');
  }
  let offset = 0;
  const tag = payload[offset]!;
  offset += 1;
  const bodyLenField = readVarUint(payload, offset);
  offset = bodyLenField.nextOffset;
  const bodyEnd = offset + bodyLenField.value;
  if (bodyEnd > payload.length) {
    throw new RangeError('message body exceeds bounds');
  }
  const body = payload.subarray(offset, bodyEnd);
  offset = bodyEnd;
  if (offset !== payload.length) {
    throw new RangeError('unexpected trailing bytes in message payload');
  }

  if (tag === KEYFRAME_MESSAGE_TAG) {
    return { kind: 'keyframe', keyframePayload: body };
  }
  if (tag === TELEPORT_MESSAGE_TAG) {
    return { kind: 'teleport', teleportPacket: deserializeTeleportPacket(body) };
  }
  throw new RangeError(`unsupported message tag ${tag}`);
}

function framePayloadOverChannels(
  payload: Uint8Array,
  config: TransportConfig
): AeonFrameChunk[] {
  const frames: AeonFrameChunk[] = [];
  for (
    let offset = 0, index = 0;
    offset < payload.length;
    offset += config.framePayloadBytes, index++
  ) {
    const end = Math.min(offset + config.framePayloadBytes, payload.length);
    frames.push({
      channel: index % config.channels,
      sequence: index,
      payload: payload.subarray(offset, end),
    });
  }
  if (frames.length === 0) {
    frames.push({
      channel: 0,
      sequence: 0,
      payload: new Uint8Array(0),
    });
  }
  return frames;
}

function shuffled<T>(items: readonly T[], seed: number): T[] {
  const rng = createLcg(seed);
  const out = [...items];
  for (let i = out.length - 1; i > 0; i--) {
    const j = Math.floor(rng() * (i + 1));
    const tmp = out[i]!;
    out[i] = out[j]!;
    out[j] = tmp;
  }
  return out;
}

function reassembleFrames(frames: readonly AeonFrameChunk[]): Uint8Array {
  const ordered = [...frames].sort((a, b) => a.sequence - b.sequence);
  return concatUint8Arrays(ordered.map((frame) => frame.payload));
}

function sendThroughAeon(
  messagePayload: Uint8Array,
  config: TransportConfig,
  seed: number
): { reassembled: Uint8Array; wireBytes: number; frameCount: number } {
  const frames = framePayloadOverChannels(messagePayload, config);
  const wireBytes =
    messagePayload.length + frames.length * AEON_HEADER_BYTES;
  const arrivalOrder = shuffled(frames, seed);
  const reassembled = reassembleFrames(arrivalOrder);
  return { reassembled, wireBytes, frameCount: frames.length };
}

function brotliCompressUtf16(text: string): Uint8Array {
  return new Uint8Array(
    brotliCompressSync(utf16leBytes(text), {
      params: {
        [zlibConstants.BROTLI_PARAM_QUALITY]: 11,
        [zlibConstants.BROTLI_PARAM_MODE]: zlibConstants.BROTLI_MODE_TEXT,
      },
    })
  );
}

function brotliDecompressUtf16(payload: Uint8Array): string {
  const out = brotliDecompressSync(Buffer.from(payload));
  return fromUtf16leBytes(new Uint8Array(out));
}

function formatPct(value: number): string {
  return `${value.toFixed(1)}%`;
}

function randomChar(rng: () => number): string {
  const alphabet = [
    ...'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789 .,;:-_',
    String.fromCodePoint(0x1f642),
    String.fromCodePoint(0x1f680),
    'λ',
    '∑',
  ];
  return alphabet[Math.floor(rng() * alphabet.length)]!;
}

function randomText(length: number, rng: () => number): string {
  let out = '';
  for (let i = 0; i < length; i++) out += randomChar(rng);
  return out;
}

function mutateText(base: string, operations: number, rng: () => number): string {
  const chars = Array.from(base);
  for (let i = 0; i < operations; i++) {
    const mode = rng();
    if (mode < 0.34 && chars.length > 1) {
      const idx = Math.floor(rng() * chars.length);
      chars.splice(idx, 1);
      continue;
    }
    if (mode < 0.67) {
      const idx = Math.floor(rng() * (chars.length + 1));
      chars.splice(idx, 0, randomChar(rng));
      continue;
    }
    const idx = Math.floor(rng() * Math.max(1, chars.length));
    chars[idx] = randomChar(rng);
  }
  return chars.join('');
}

function buildScenarios(): Scenario[] {
  const scenarios: Scenario[] = [];

  {
    const states: string[] = [];
    let current = 'alpha bravo charlie '.repeat(36);
    const rng = createLcg(0x41cafe);
    for (let t = 0; t < 48; t++) {
      current = mutateText(current, 2, rng);
      states.push(current);
    }
    scenarios.push({ name: 'baseline-close-edit-stream', states });
  }

  {
    const states: string[] = [];
    const rng = createLcg(0x99b00b);
    let current = randomText(2400, rng);
    for (let t = 0; t < 56; t++) {
      current = mutateText(current, 18, rng);
      states.push(current);
    }
    scenarios.push({ name: 'collaborative-drift-stream', states });
  }

  {
    const states: string[] = [];
    const rng = createLcg(0xdeadbeef);
    let current = randomText(2800, rng);
    for (let t = 0; t < 40; t++) {
      if (t % 5 === 0 && t > 0) {
        current = randomText(2800, rng);
      } else {
        current = mutateText(current, 25, rng);
      }
      states.push(current);
    }
    scenarios.push({ name: 'periodic-rewrite-stream', states });
  }

  return scenarios;
}

function defaultStrategies(): StrategyConfig[] {
  return [
    { name: 'brotli-full', kind: 'brotli-full' },
    { name: 'brotli-full-utf8', kind: 'brotli-full-utf8' },
    {
      name: 'unspeech-residual-k8',
      kind: 'unspeech-hybrid',
      keyframeInterval: 8,
      teleportMode: 'residual',
      reorderMode: 'none',
    },
    {
      name: 'unspeech-auto-k8',
      kind: 'unspeech-hybrid',
      keyframeInterval: 8,
      teleportMode: 'auto',
      reorderMode: 'auto',
    },
    {
      name: 'unspeech-residual-no-reset',
      kind: 'unspeech-hybrid',
      keyframeInterval: Number.MAX_SAFE_INTEGER,
      teleportMode: 'residual',
      reorderMode: 'none',
    },
  ];
}

async function evaluateStrategy(
  scenario: Scenario,
  strategy: StrategyConfig,
  transport: TransportConfig
): Promise<StrategyResult> {
  let rawUtf16Bytes = 0;
  let wireBytes = 0;
  let frameCount = 0;
  const packetKinds: Record<string, number> = {};

  let senderState = '';
  let receiverState = '';
  const lexicon = buildUtf16Lexicon(scenario.states);

  for (let tick = 0; tick < scenario.states.length; tick++) {
    const targetState = scenario.states[tick]!;
    rawUtf16Bytes += utf16leBytes(targetState).length;

    let wirePayload: Uint8Array;
    let expectedKind: 'keyframe' | 'teleport';

    if (strategy.kind === 'brotli-full') {
      wirePayload = serializeKeyframeMessage(brotliCompressUtf16(targetState));
      expectedKind = 'keyframe';
    } else if (strategy.kind === 'brotli-full-utf8') {
      wirePayload = serializeKeyframeMessage(
        new Uint8Array(
          brotliCompressSync(utf8Bytes(targetState), {
            params: {
              [zlibConstants.BROTLI_PARAM_QUALITY]: 11,
              [zlibConstants.BROTLI_PARAM_MODE]: zlibConstants.BROTLI_MODE_TEXT,
            },
          })
        )
      );
      expectedKind = 'keyframe';
    } else {
      const interval = strategy.keyframeInterval ?? 8;
      const isKeyframe = tick === 0 || (interval > 0 && tick % interval === 0);

      if (isKeyframe) {
        wirePayload = serializeKeyframeMessage(brotliCompressUtf16(targetState));
        expectedKind = 'keyframe';
      } else {
        const packet = await teleportUtf16ThroughLexicon(targetState, lexicon, {
          baselineText: senderState,
          mode: strategy.teleportMode ?? 'residual',
          reorderMode: strategy.reorderMode ?? 'none',
          tritonCompression: {
            motifSize: 5,
            useBitwiseFold: false,
            preferSmallerPayload: true,
          },
        });
        packetKinds[packet.strategy] = (packetKinds[packet.strategy] ?? 0) + 1;
        wirePayload = serializeTeleportMessage(packet);
        expectedKind = 'teleport';
      }
    }

    const transmitted = sendThroughAeon(
      wirePayload,
      transport,
      (tick + 1) * 17 + scenario.states.length
    );
    wireBytes += transmitted.wireBytes;
    frameCount += transmitted.frameCount;

    const decoded = deserializeMessage(transmitted.reassembled);
    if (decoded.kind !== expectedKind) {
      throw new Error(
        `${scenario.name}/${strategy.name}: expected ${expectedKind} but received ${decoded.kind} at tick ${tick}`
      );
    }

    let reconstructed: string;
    if (decoded.kind === 'keyframe') {
      if (strategy.kind === 'brotli-full-utf8') {
        reconstructed = fromUtf8Bytes(
          new Uint8Array(brotliDecompressSync(Buffer.from(decoded.keyframePayload!)))
        );
      } else {
        reconstructed = brotliDecompressUtf16(decoded.keyframePayload!);
      }
      packetKinds.keyframe = (packetKinds.keyframe ?? 0) + 1;
    } else {
      reconstructed = await inflateUtf16FromTeleport(
        decoded.teleportPacket!,
        lexicon,
        { baselineText: receiverState }
      );
    }

    if (reconstructed !== targetState) {
      throw new Error(
        `${scenario.name}/${strategy.name}: state mismatch at tick ${tick}`
      );
    }

    senderState = targetState;
    receiverState = reconstructed;
  }

  return {
    strategy: strategy.name,
    rawUtf16Bytes,
    wireBytes,
    frameCount,
    avgFramesPerTick: frameCount / Math.max(1, scenario.states.length),
    compressionSavingsPct:
      rawUtf16Bytes === 0 ? 0 : (1 - wireBytes / rawUtf16Bytes) * 100,
    packetKinds,
  };
}

async function runScenario(
  scenario: Scenario,
  strategies: readonly StrategyConfig[],
  transport: TransportConfig
): Promise<ScenarioResult> {
  const results: StrategyResult[] = [];
  for (const strategy of strategies) {
    results.push(await evaluateStrategy(scenario, strategy, transport));
  }
  const winner = [...results].sort((a, b) => a.wireBytes - b.wireBytes)[0]!;
  return {
    scenario: scenario.name,
    tickCount: scenario.states.length,
    results,
    winner,
  };
}

function printScenario(result: ScenarioResult): void {
  console.log(`scenario: ${result.scenario} (ticks=${result.tickCount})`);
  for (const stat of result.results) {
    const mix = Object.entries(stat.packetKinds)
      .map(([kind, count]) => `${kind}:${count}`)
      .join(', ');
    console.log(
      `  ${stat.strategy.padEnd(26)} ${String(stat.wireBytes).padStart(
        8
      )}B  savings=${formatPct(stat.compressionSavingsPct).padStart(
        7
      )}  frames=${String(stat.frameCount).padStart(4)}  avgFrames/tick=${stat.avgFramesPerTick.toFixed(
        2
      )}${mix ? `  mix=[${mix}]` : ''}`
    );
  }
  console.log(
    `  winner: ${result.winner.strategy} (${result.winner.wireBytes}B)`
  );
  console.log('');
}

/**
 * Runs the gnosis command-line workflow.
 */
export async function main(): Promise<void> {
  const transport: TransportConfig = {
    channels: 4,
    framePayloadBytes: 1024,
  };
  const scenarios = buildScenarios();
  const strategies = defaultStrategies();

  console.log(
    `aeon transport config: channels=${transport.channels}, framePayloadBytes=${transport.framePayloadBytes}, headerBytes=${AEON_HEADER_BYTES}`
  );
  console.log('');

  for (const scenario of scenarios) {
    const result = await runScenario(scenario, strategies, transport);
    printScenario(result);
  }
}

const invokedPath = process.argv[1];
if (invokedPath && import.meta.url === pathToFileURL(invokedPath).href) {
  main().catch((error: unknown) => {
    console.error(error);
    process.exit(1);
  });
}
