// @a0n/.ignored_gnosis/src/benchmarks/ambient-substrate-benchmark.ts
import { performance } from 'node:perf_hooks';
import { pathToFileURL } from 'node:url';
import {
  GNOSIS_AMBIENT_SUBSTRATES,
  gnosisAmbientSubstrateEnv,
  gnosisAmbientSubstrateWorkIndex,
  planGnosisAmbientSubstrateSchedule,
  type GnosisAmbientSubstrate,
} from '../runtime/ambient-substrate';
import {
  bootstrapMeanConfidenceInterval,
  mean,
  stdev,
  type ConfidenceInterval,
} from './statistics';

export interface AmbientSubstrateBenchmarkOptions {
  readonly iterations?: number;
  readonly warmup?: number;
  readonly workItems?: number;
  readonly memoryBudgetBytes?: number;
  readonly bytesPerItem?: number;
  readonly maxParallelism?: number;
}

export interface AmbientSubstrateBenchmarkRow {
  readonly substrate: GnosisAmbientSubstrate;
  readonly iterations: number;
  readonly workItems: number;
  readonly checksum: number;
  readonly scheduleMeanNs: number;
  readonly indexMeanNs: number;
  readonly envMeanNs: number;
  readonly scheduleOpsPerSec: number;
  readonly indexOpsPerSec: number;
  readonly envOpsPerSec: number;
  readonly scheduleStdevNs: number;
  readonly indexStdevNs: number;
  readonly envStdevNs: number;
  readonly scheduleCi95Ns: ConfidenceInterval;
  readonly indexCi95Ns: ConfidenceInterval;
  readonly envCi95Ns: ConfidenceInterval;
}

export interface AmbientSubstrateBenchmarkReport {
  readonly label: 'gnosis-ambient-substrate-benchmark-v1';
  readonly iterations: number;
  readonly warmup: number;
  readonly workItems: number;
  readonly memoryBudgetBytes: number;
  readonly bytesPerItem: number;
  readonly maxParallelism: number;
  readonly substrates: readonly AmbientSubstrateBenchmarkRow[];
}

interface NormalizedAmbientSubstrateBenchmarkOptions {
  readonly iterations: number;
  readonly warmup: number;
  readonly workItems: number;
  readonly memoryBudgetBytes: number;
  readonly bytesPerItem: number;
  readonly maxParallelism: number;
}

/**
 * Handles the gnosis run Ambient Substrate Benchmark workflow.
 */
export function runAmbientSubstrateBenchmark(
  options: AmbientSubstrateBenchmarkOptions = {}
): AmbientSubstrateBenchmarkReport {
  const normalized = normalizeOptions(options);
  const rows = GNOSIS_AMBIENT_SUBSTRATES.map((substrate) =>
    benchmarkSubstrate(substrate, normalized)
  );

  return {
    label: 'gnosis-ambient-substrate-benchmark-v1',
    ...normalized,
    substrates: rows,
  };
}

function benchmarkSubstrate(
  substrate: GnosisAmbientSubstrate,
  options: NormalizedAmbientSubstrateBenchmarkOptions
): AmbientSubstrateBenchmarkRow {
  for (let i = 0; i < options.warmup; i++) {
    const schedule = makeSchedule(substrate, options, i);
    gnosisAmbientSubstrateWorkIndex(schedule, i, options.workItems);
    gnosisAmbientSubstrateEnv(schedule);
  }

  const scheduleNs: number[] = [];
  const indexNs: number[] = [];
  const envNs: number[] = [];
  let checksum = 0;
  let lastSchedule = makeSchedule(substrate, options, 0);

  for (let i = 0; i < options.iterations; i++) {
    let started = performance.now();
    lastSchedule = makeSchedule(substrate, options, i);
    scheduleNs.push(toNanoseconds(performance.now() - started));

    started = performance.now();
    checksum ^= gnosisAmbientSubstrateWorkIndex(
      lastSchedule,
      i,
      options.workItems
    );
    indexNs.push(toNanoseconds(performance.now() - started));

    started = performance.now();
    checksum ^= Number(
      gnosisAmbientSubstrateEnv(lastSchedule).GNOSIS_AMBIENT_PARALLELISM
    );
    envNs.push(toNanoseconds(performance.now() - started));
  }

  const scheduleMeanNs = mean(scheduleNs);
  const indexMeanNs = mean(indexNs);
  const envMeanNs = mean(envNs);

  return {
    substrate,
    iterations: options.iterations,
    workItems: options.workItems,
    checksum,
    scheduleMeanNs,
    indexMeanNs,
    envMeanNs,
    scheduleOpsPerSec: opsPerSecond(scheduleMeanNs),
    indexOpsPerSec: opsPerSecond(indexMeanNs),
    envOpsPerSec: opsPerSecond(envMeanNs),
    scheduleStdevNs: stdev(scheduleNs),
    indexStdevNs: stdev(indexNs),
    envStdevNs: stdev(envNs),
    scheduleCi95Ns: bootstrapMeanConfidenceInterval(scheduleNs),
    indexCi95Ns: bootstrapMeanConfidenceInterval(indexNs),
    envCi95Ns: bootstrapMeanConfidenceInterval(envNs),
  };
}

function makeSchedule(
  substrate: GnosisAmbientSubstrate,
  options: NormalizedAmbientSubstrateBenchmarkOptions,
  iteration: number
) {
  return planGnosisAmbientSubstrateSchedule({
    substrate,
    foldedBytes: Uint8Array.from([
      iteration & 0xff,
      (iteration >>> 8) & 0xff,
      7,
      10,
      13,
      21,
      34,
      55,
    ]),
    cfRay: `${iteration.toString(16).padStart(16, '0')}-SJC`,
    colo: 'SJC',
    clientTcpRtt: 11 + (iteration % 31),
    timestampMs: 1_765_000_000_000 + iteration,
    workItems: options.workItems,
    memoryBudgetBytes: options.memoryBudgetBytes,
    bytesPerItem: options.bytesPerItem,
    maxParallelism: options.maxParallelism,
  });
}

function normalizeOptions(
  options: AmbientSubstrateBenchmarkOptions
): NormalizedAmbientSubstrateBenchmarkOptions {
  return {
    iterations: positiveInteger(options.iterations ?? 50_000, 'iterations'),
    warmup: positiveInteger(options.warmup ?? 2_000, 'warmup'),
    workItems: positiveInteger(options.workItems ?? 1024, 'workItems'),
    memoryBudgetBytes: positiveInteger(
      options.memoryBudgetBytes ?? 1 << 20,
      'memoryBudgetBytes'
    ),
    bytesPerItem: positiveInteger(options.bytesPerItem ?? 256, 'bytesPerItem'),
    maxParallelism: positiveInteger(
      options.maxParallelism ?? 64,
      'maxParallelism'
    ),
  };
}

function parseNumericFlag(
  argv: readonly string[],
  name: string,
  fallback: number
): number {
  const prefix = `--${name}=`;
  const value = argv.find((arg) => arg.startsWith(prefix))?.slice(prefix.length);
  return value === undefined ? fallback : Number(value);
}

function positiveInteger(value: number, name: string): number {
  if (!Number.isInteger(value) || value < 1) {
    throw new RangeError(`${name} must be a positive integer`);
  }
  return value;
}

function toNanoseconds(milliseconds: number): number {
  return milliseconds * 1_000_000;
}

function opsPerSecond(meanNs: number): number {
  return meanNs === 0 ? 0 : 1_000_000_000 / meanNs;
}

/**
 * Runs the gnosis command-line workflow.
 */
export async function main(): Promise<AmbientSubstrateBenchmarkReport> {
  const argv = process.argv.slice(2);
  const report = runAmbientSubstrateBenchmark({
    iterations: parseNumericFlag(argv, 'iterations', 50_000),
    warmup: parseNumericFlag(argv, 'warmup', 2_000),
    workItems: parseNumericFlag(argv, 'workItems', 1024),
    memoryBudgetBytes: parseNumericFlag(argv, 'memoryBudgetBytes', 1 << 20),
    bytesPerItem: parseNumericFlag(argv, 'bytesPerItem', 256),
    maxParallelism: parseNumericFlag(argv, 'maxParallelism', 64),
  });
  console.log(JSON.stringify(report, null, 2));
  return report;
}

const invokedPath = process.argv[1];
if (invokedPath && import.meta.url === pathToFileURL(invokedPath).href) {
  main().catch((error: unknown) => {
    console.error(error);
    process.exitCode = 1;
  });
}
