// @a0n/.ignored_gnosis/src/benchmarks/dark-matvec-benchmark.ts
#!/usr/bin/env gnode
/**
 * dark-matvec-benchmark.ts -- Gnosis Dark Fork MatVec vs Normal Chunked MatVec
 *
 * Balanced Equations for stability (no beachball of death).
 * Chunks: 2048 | Classical: 1024 | Dark: 1024 | Fabric Limit: 4096
 *
 * CLI (all optional):
 *   --quick              Tiny workload (default for a fast smoke run when passed)
 *   --total-chunks=N     Total chunk count (default 2048)
 *   --chunk-size=N       Inner-loop work per chunk (default 500)
 *   --classical-limit=N  Classical mesh cap (default 1024)
 *   --drain-timeout-ms=N Max wait for dark queue to drain (default 120000)
 *   --matrix             Full grid: totalChunks × chunkSize (classical = half of totalChunks per cell)
 */

import { UnboundedDarkEnergyFork, safeFabricDimensionBudgetForDarkChunks } from '../dark-fork-topology';
import { createEngine, type Task } from '../mesh-engine';

// ═══════════════════════════════════════════════════════════════════════════════
// Mock MatVec Operations
// ═══════════════════════════════════════════════════════════════════════════════

function computeMatVecChunk(chunkSize: number): number {
  let sum = 0;
  // Reduce internal complexity slightly to keep event loop healthy
  for (let i = 0; i < chunkSize; i++) {
    sum += Math.random() * 0.5;
  }
  return sum;
}

export interface MatVecBenchConfig {
  totalChunks: number;
  chunkSize: number;
  classicalWorkerLimit: number;
  drainTimeoutMs: number;
}

const DEFAULT_CONFIG: MatVecBenchConfig = {
  totalChunks: 2048,
  chunkSize: 500,
  classicalWorkerLimit: 1024,
  drainTimeoutMs: 120_000,
};

/** Minimal workload for a fast sanity check (low bytes / low iteration count). */
const QUICK_CONFIG: MatVecBenchConfig = {
  totalChunks: 48,
  chunkSize: 32,
  classicalWorkerLimit: 24,
  drainTimeoutMs: 30_000,
};

/** Rows (totalChunks) × columns (chunkSize) for `--matrix`. Classical limit = floor(totalChunks/2) per cell. */
const MATRIX_TOTAL_CHUNKS_AXIS: readonly number[] = [64, 256, 1024, 2048];
const MATRIX_CHUNK_SIZE_AXIS: readonly number[] = [32, 128, 500];
/** Per-cell drain cap so `--matrix` finishes; large rows may still show STALL. */
const MATRIX_CELL_DRAIN_TIMEOUT_MS = 15_000;

function parseMatVecBenchConfig(argv: string[]): MatVecBenchConfig {
  let cfg: MatVecBenchConfig = { ...DEFAULT_CONFIG };

  for (const raw of argv) {
    if (raw === '--quick') {
      cfg = { ...QUICK_CONFIG };
      continue;
    }
    if (raw === '--matrix') {
      continue;
    }
    if (raw === '--help' || raw === '-h') {
      console.log(`Usage: gnode run dark-matvec-benchmark.ts --export main -- [options]
Options:
  --quick                 Use tiny preset (${QUICK_CONFIG.totalChunks} chunks, chunk-size ${QUICK_CONFIG.chunkSize}, classical ${QUICK_CONFIG.classicalWorkerLimit})
  --matrix                Sweep totalChunks × chunkSize grid (see source for axes)
  --total-chunks=N
  --chunk-size=N
  --classical-limit=N
  --drain-timeout-ms=N`);
      process.exit(0);
    }
    const eq = raw.indexOf('=');
    if (eq <= 0) continue;
    const key = raw.slice(0, eq);
    const val = Number(raw.slice(eq + 1));
    if (!Number.isFinite(val) || val < 0) continue;
    if (key === '--total-chunks') cfg.totalChunks = Math.floor(val);
    if (key === '--chunk-size') cfg.chunkSize = Math.floor(val);
    if (key === '--classical-limit') cfg.classicalWorkerLimit = Math.floor(val);
    if (key === '--drain-timeout-ms') cfg.drainTimeoutMs = Math.floor(val);
  }

  if (cfg.classicalWorkerLimit > cfg.totalChunks) {
    cfg.classicalWorkerLimit = cfg.totalChunks;
  }
  if (cfg.totalChunks < 1) cfg.totalChunks = 1;
  if (cfg.chunkSize < 1) cfg.chunkSize = 1;
  if (cfg.classicalWorkerLimit < 0) cfg.classicalWorkerLimit = 0;

  return cfg;
}

/** Args after a lone `--` (for `gnode run file -- --flags`); otherwise full argv slice. */
function argvAfterDoubleDash(argv: readonly string[]): string[] {
  const idx = argv.indexOf('--');
  if (idx < 0) {
    return [...argv];
  }
  return argv.slice(idx + 1);
}

function padLeft(s: string, width: number): string {
  if (s.length >= width) {
    return s;
  }
  return ' '.repeat(width - s.length) + s;
}

function padCellMs(ms: number, width: number): string {
  const t = Number.isFinite(ms) ? ms.toFixed(2) : String(ms);
  return padLeft(t, width);
}

async function runFullMatrix(drainTimeoutMs: number): Promise<void> {
  const cellDrain = Math.min(drainTimeoutMs, MATRIX_CELL_DRAIN_TIMEOUT_MS);
  const colW = 12;
  const labelW = 10;
  const colHead =
    padLeft('T \\ S', labelW) +
    MATRIX_CHUNK_SIZE_AXIS.map((c) => padLeft(String(c), colW)).join('');

  const normalLines: string[] = [];
  const hybridLines: string[] = [];
  const drainLines: string[] = [];

  for (const totalChunks of MATRIX_TOTAL_CHUNKS_AXIS) {
    const classicalWorkerLimit = Math.floor(totalChunks / 2);
    const rowN = padLeft(String(totalChunks), labelW);
    const rowH = padLeft(String(totalChunks), labelW);
    const rowD = padLeft(String(totalChunks), labelW);
    const cellsN: string[] = [];
    const cellsH: string[] = [];
    const cellsD: string[] = [];

    for (const chunkSize of MATRIX_CHUNK_SIZE_AXIS) {
      const cfg: MatVecBenchConfig = {
        totalChunks,
        chunkSize,
        classicalWorkerLimit,
        drainTimeoutMs: cellDrain,
      };
      process.stderr.write(`[matrix] T=${totalChunks} S=${chunkSize} classical=${classicalWorkerLimit} …\n`);
      const n = await runNormalMeshLimited(cfg, false);
      const h = await runHybridDarkMatVec(cfg, false);
      cellsN.push(padCellMs(n.elapsed, colW));
      cellsH.push(padCellMs(h.elapsed, colW));
      cellsD.push(padLeft(h.darkDrained ? 'ok' : 'STALL', colW));
    }
    normalLines.push(rowN + cellsN.join(''));
    hybridLines.push(rowH + cellsH.join(''));
    drainLines.push(rowD + cellsD.join(''));
  }

  console.log('═══════════════════════════════════════════════════════════════');
  console.log(' Dark-MatVec FULL MATRIX');
  console.log(
    ' Rows: totalChunks in ' + MATRIX_TOTAL_CHUNKS_AXIS.join(', ') + ' | Cols: chunkSize ' + MATRIX_CHUNK_SIZE_AXIS.join(', ')
  );
  console.log(' classicalWorkerLimit = floor(totalChunks/2) per row | Times in ms');
  console.log(
    ' Per-cell drainTimeoutMs = min(cli,' + String(MATRIX_CELL_DRAIN_TIMEOUT_MS) + ') = ' + String(cellDrain)
  );
  console.log('═══════════════════════════════════════════════════════════════');

  console.log('\n--- Normal mesh (strict classical cap): wall time (ms) ---\n');
  console.log(colHead);
  for (const line of normalLines) {
    console.log(line);
  }

  console.log('\n--- Hybrid dark MatVec: wall time (ms) ---\n');
  console.log(colHead);
  for (const line of hybridLines) {
    console.log(line);
  }

  console.log('\n--- Hybrid drain status (expected dark chunks completed) ---\n');
  console.log(colHead);
  for (const line of drainLines) {
    console.log(line);
  }
  console.log('');
}

// ═══════════════════════════════════════════════════════════════════════════════
// Benchmarks
// ═══════════════════════════════════════════════════════════════════════════════

async function runNormalMeshLimited(cfg: MatVecBenchConfig, verbose = true) {
  if (verbose) {
    console.log('\n── Scenario 1: Normal Mesh (Strict Limit) ──────────────');
  }
  const start = performance.now();

  const promises: Promise<number>[] = [];
  let failedChunks = 0;

  for (let i = 0; i < cfg.totalChunks; i++) {
    if (i < cfg.classicalWorkerLimit) {
      promises.push(
        new Promise<number>((resolve) => {
          // Yield to event loop
          setImmediate(() => resolve(computeMatVecChunk(cfg.chunkSize)));
        })
      );
    } else {
      failedChunks++;
      promises.push(Promise.resolve(0));
    }
  }

  const results = await Promise.all(promises);
  const totalSum = results.reduce((acc, val) => acc + val, 0);
  const elapsed = performance.now() - start;

  if (verbose) {
    console.log(`Requested Chunks: ${cfg.totalChunks}`);
    console.log(`Computed Chunks:  ${cfg.totalChunks - failedChunks}`);
    console.log(`Failed/Dropped:   ${failedChunks}`);
    console.log(`Time:             ${elapsed.toFixed(2)}ms`);
  }

  return { sum: totalSum, elapsed };
}

async function runHybridDarkMatVec(cfg: MatVecBenchConfig, verbose = true) {
  if (verbose) {
    console.log('\n── Scenario 2: Hybrid Dark MatVec (Balanced) ────────────────');
  }

  const engine = createEngine('syzygy', 4); // Use a slightly larger mesh

  const start = performance.now();
  const normalPromises: Promise<number>[] = [];

  let darkQueueSize = 0;
  const darkResults: number[] = [];

  for (let i = 0; i < cfg.totalChunks; i++) {
    if (i < cfg.classicalWorkerLimit) {
      normalPromises.push(
        new Promise<number>((resolve) => {
          setImmediate(() => resolve(computeMatVecChunk(cfg.chunkSize)));
        })
      );
    } else {
      darkQueueSize++;
    }
  }

  const expectedDark = Math.max(0, cfg.totalChunks - cfg.classicalWorkerLimit);
  // Balanced Delay: 1ms between expansions; fabric cap scales with expected dark chunks (per-request budget)
  const fabricBudget = safeFabricDimensionBudgetForDarkChunks(expectedDark, 4);
  const darkEnergy = new UnboundedDarkEnergyFork(engine, 1, fabricBudget);

  // Dark Task: Processes batch of 4 per "expansion" tick (must match budget helper batch size)
  const darkTask: Task<void> = async (_vertex, _signal) => {
    for (let b = 0; b < 4; b++) {
      if (darkQueueSize > 0) {
        darkQueueSize--;
        const result = computeMatVecChunk(cfg.chunkSize);
        darkResults.push(result);
      }
    }
  };

  darkEnergy.start(darkTask);

  const normalResults = await Promise.all(normalPromises);
  let totalSum = normalResults.reduce((acc, val) => acc + val, 0);

  const drainDeadline = performance.now() + cfg.drainTimeoutMs;
  while (darkQueueSize > 0) {
    if (performance.now() > drainDeadline) {
      if (verbose) {
        console.error(
          `[dark-matvec-benchmark] Dark queue did not drain within ${cfg.drainTimeoutMs}ms (remaining=${darkQueueSize}). ` +
            `The dark fork may have stalled (e.g. fabric safety stop). Try --quick or smaller --total-chunks.`
        );
      }
      break;
    }
    await new Promise((resolve) => setTimeout(resolve, 20));
  }
  darkEnergy.stop();

  totalSum += darkResults.reduce((acc, val) => acc + val, 0);
  const elapsed = performance.now() - start;

  const darkDrained = darkQueueSize === 0 && darkResults.length === expectedDark;

  if (verbose) {
    console.log(`Requested Chunks: ${cfg.totalChunks}`);
    console.log(`Normal Computed:  ${cfg.classicalWorkerLimit}`);
    console.log(`Dark Computed:    ${darkResults.length}`);
    console.log(`Final Sum:        ${totalSum.toFixed(2)} (Complete)`);
    console.log(`Time:             ${elapsed.toFixed(2)}ms`);
  }

  return { sum: totalSum, elapsed, darkDrained };
}

async function runMatVecBenchmark(): Promise<void> {
  const benchArgv = process.argv.slice(2);
  const forwarded = argvAfterDoubleDash(benchArgv);
  const wantMatrix = forwarded.includes('--matrix');
  const cfgArgv = forwarded.filter((a) => a !== '--matrix');
  const cfg = parseMatVecBenchConfig(cfgArgv);

  if (wantMatrix) {
    await runFullMatrix(cfg.drainTimeoutMs);
    return;
  }

  console.log('═══════════════════════════════════════════════════════════════');
  console.log(' Hybrid Dark MatVec vs Strict Limit Benchmark (Balanced)');
  console.log('═══════════════════════════════════════════════════════════════');
  console.log(
    `Config: totalChunks=${cfg.totalChunks} chunkSize=${cfg.chunkSize} classicalLimit=${cfg.classicalWorkerLimit} drainTimeoutMs=${cfg.drainTimeoutMs}`
  );

  console.log('\n── Reference run ────────────────────────────────────────────');
  await runNormalMeshLimited(cfg);
  const hybrid = await runHybridDarkMatVec(cfg);

  console.log('\nInterpretation:');
  console.log(` - Normal (Limited): Capped at ${cfg.classicalWorkerLimit}. Result is truncated when totalChunks exceeds classical limit.`);
  console.log(` - Hybrid Dark: Balanced at ${hybrid.elapsed.toFixed(2)}ms.`);
  console.log(`   Correctness achieved via Hilbert space overflow handling.`);
}

/**
 * Runs the gnosis command-line workflow.
 */
export async function main() {
  await runMatVecBenchmark();
  return;
}

void main().catch(console.error);
