// @a0n/.ignored_gnosis/src/benchmarks/ehla-performance-bench.ts
import { FoilEHLARuntime } from '../foil-ehla-runtime';
import { 
  ZERO_DEFICIT_FOIL_CERTIFICATE, 
  createStaticFoilAdmissionProof 
} from '../foil-edge-gate';
import { performance } from 'node:perf_hooks';

/**
 * EHLA Performance Benchmark
 * 
 * Measures the throughput and latency improvements of EHLA (Entropy-Harvesting 
 * Latent Architecture) compared to Rigid (blocking) execution.
 * 
 * EHLA speedup comes from harvesting the "interpretation gap" while 
 * proofs are verified latently.
 */

async function sleep(ms: number) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

// A verifier that simulates a heavy proof verification (e.g. Lean 4 static proving)
const heavyVerifier = (proof: any) => {
  // Simulate 20ms of work per verification
  const start = performance.now();
  while (performance.now() - start < 20) {
    // busy loop
  }
  return { accepted: true, reason: 'verified' };
};

/**
 * Handles the gnosis run EHLABenchmark workflow.
 */
export async function runEHLABenchmark() {
  const iterations = 50;
  const maxDebt = 10;
  
  console.log(`
[EHLA Benchmark] Starting shootout: ${iterations} iterations, maxDebt=${maxDebt}`);
  console.log(`[EHLA Benchmark] Simulated Verification Cost: 20ms/exec`);

  const payload = {
    ggGraph: new Uint8Array([1]),
    zkProof: createStaticFoilAdmissionProof(ZERO_DEFICIT_FOIL_CERTIFICATE)
  };

  // --- 1. RIGID MODE (Blocking) ---
  const rigidRuntime = new FoilEHLARuntime({ 
    harvestingEnabled: false, 
    verifier: heavyVerifier 
  });
  
  const startRigid = performance.now();
  for (let i = 0; i < iterations; i++) {
    rigidRuntime.executeTopologyEHLA(payload, () => {
      // Simulate minimal useful work
    });
  }
  const endRigid = performance.now();
  const rigidTime = endRigid - startRigid;
  const rigidThroughput = (iterations / (rigidTime / 1000)).toFixed(2);

  console.log(`[RIGID] Total Time: ${rigidTime.toFixed(2)}ms`);
  console.log(`[RIGID] Throughput: ${rigidThroughput} exec/s`);

  // --- 2. EHLA MODE (Harvesting) ---
  // In a real system, debt resolution happens in the background.
  // Here we simulate the harvester by resolving debt every few calls.
  const ehlaRuntime = new FoilEHLARuntime({ 
    harvestingEnabled: true, 
    maxDebt, 
    verifier: heavyVerifier 
  });
  
  const startEHLA = performance.now();
  for (let i = 0; i < iterations; i++) {
    ehlaRuntime.executeTopologyEHLA(payload, () => {
      // Simulate minimal useful work
    });
    
    // Background harvester thread: resolves 1 debt unit after it was "latently" verified
    if (i % 2 === 0) {
        ehlaRuntime.resolveDebt(1);
    }
  }
  const endEHLA = performance.now();
  const ehlaTime = endEHLA - startEHLA;
  const ehlaThroughput = (iterations / (ehlaTime / 1000)).toFixed(2);
  const speedup = (rigidTime / ehlaTime).toFixed(2);

  console.log(`[EHLA]  Total Time: ${ehlaTime.toFixed(2)}ms`);
  console.log(`[EHLA]  Throughput: ${ehlaThroughput} exec/s`);
  console.log(`
[RESULT] EHLA Speedup: ${speedup}x`);
  
  const stats = ehlaRuntime.getEHLAStats();
  console.log(`[STATS]  Harvested: ${stats.harvestedTopologies}/${iterations}`);
  console.log(`[STATS]  Verifier Calls Avoided: ${iterations - stats.verifierCalls}`);

  return {
    rigidTime,
    ehlaTime,
    speedup,
    stats
  };
}

if (import.meta.main || (typeof process !== 'undefined' && process.argv[1] === fileURLToPath(import.meta.url))) {
    runEHLABenchmark();
}

function fileURLToPath(url: string): string {
    return url.replace('file://', '');
}
