// @a0n/.ignored_gnosis/src/benchmarks/expressiveness-benchmark.ts
#!/usr/bin/env bun
/**
 * Expressiveness Benchmark -- Code Golf for Distributed Systems
 *
 * Compares gnosis .gg topologies against equivalent solutions in
 * Go, Rust, Erlang, and TypeScript for common distributed systems
 * problems.
 *
 * Problems:
 *   1. Leader Election
 *   2. Consensus (simple majority)
 *   3. Load Balancer (round-robin with health checks)
 *   4. Circuit Breaker
 *   5. Saga Pattern (with compensation)
 *   6. CRDT Convergence (G-Counter)
 *
 * Metrics:
 *   - Lines of code (LOC)
 *   - Structural complexity (Betti number beta_1)
 *   - Primitives used (fork, race, fold, vent, process)
 *
 * Usage:
 *   bun run src/benchmarks/expressiveness-benchmark.ts
 */

// ═══════════════════════════════════════════════════════════════════════════════
// Types
// ═══════════════════════════════════════════════════════════════════════════════

interface Solution {
  readonly language: string;
  readonly loc: number; // lines of code (non-blank, non-comment)
  readonly source: string;
}

interface Problem {
  readonly name: string;
  readonly description: string;
  readonly gnosis: Solution;
  readonly alternatives: readonly Solution[];
  readonly gnosisPrimitives: readonly string[];
  readonly beta1: number; // structural Betti number
}

// ═══════════════════════════════════════════════════════════════════════════════
// Problem 1: Leader Election
// ═══════════════════════════════════════════════════════════════════════════════

const leaderElection: Problem = {
  name: 'Leader Election',
  description:
    'N candidates race to claim leadership. First to succeed wins, others vent.',
  gnosis: {
    language: 'gnosis',
    loc: 5,
    source: `(candidates: Peers { count: 5 })
(candidates)-[:FORK { fan: 'peers' }]->(candidate: Candidate)
(candidate)-[:PROCESS { op: 'claim_leadership' }]->(claim: Claim)
(claim)-[:RACE { take: 'first', failure: 'vent' }]->(leader: Leader)
(leader)-[:PROCESS { op: 'announce' }]->(done: Elected)`,
  },
  alternatives: [
    {
      language: 'Go',
      loc: 35,
      source: `// Requires channels, select, goroutines, sync.Once`,
    },
    {
      language: 'Rust',
      loc: 42,
      source: `// Requires tokio::select!, Arc<Mutex>, JoinSet`,
    },
    {
      language: 'Erlang',
      loc: 18,
      source: `// Naturally concurrent but still needs receive/after patterns`,
    },
    {
      language: 'TypeScript',
      loc: 22,
      source: `// Promise.race + AbortController + manual cleanup`,
    },
  ],
  gnosisPrimitives: ['FORK', 'PROCESS', 'RACE', 'VENT'],
  beta1: 1, // one independent cycle (race collapses fork)
};

// ═══════════════════════════════════════════════════════════════════════════════
// Problem 2: Consensus (Simple Majority)
// ═══════════════════════════════════════════════════════════════════════════════

const consensus: Problem = {
  name: 'Consensus',
  description:
    'N voters propose values, fold to majority. Vent if no majority.',
  gnosis: {
    language: 'gnosis',
    loc: 6,
    source: `(proposal: Value { value: 42 })
(proposal)-[:FORK { fan: 'voters', count: 7 }]->(voter: Voter)
(voter)-[:PROCESS { op: 'vote' }]->(vote: Vote)
(vote)-[:FOLD { strategy: 'majority', quorum: 4 }]->(result: Consensus)
(result)-[:RACE { condition: 'has_majority' }]->(accepted: Accepted)
(result)-[:VENT { condition: 'no_majority' }]->(rejected: NoConsensus)`,
  },
  alternatives: [
    {
      language: 'Go',
      loc: 48,
      source: `// Channels, WaitGroup, atomic counter, threshold check`,
    },
    {
      language: 'Rust',
      loc: 55,
      source: `// Arc<AtomicUsize>, JoinSet, tokio::sync::watch`,
    },
    {
      language: 'Erlang',
      loc: 22,
      source: `// Naturally fits but requires manual vote counting`,
    },
    {
      language: 'TypeScript',
      loc: 28,
      source: `// Promise.all + reduce + threshold check`,
    },
  ],
  gnosisPrimitives: ['FORK', 'PROCESS', 'FOLD', 'RACE', 'VENT'],
  beta1: 2, // fork → fold + conditional race/vent
};

// ═══════════════════════════════════════════════════════════════════════════════
// Problem 3: Load Balancer
// ═══════════════════════════════════════════════════════════════════════════════

const loadBalancer: Problem = {
  name: 'Load Balancer',
  description:
    'Route requests across backends with health checks. Unhealthy backends vent.',
  gnosis: {
    language: 'gnosis',
    loc: 7,
    source: `(request: HttpRequest)
(backends: Pool { hosts: ['a:80','b:80','c:80'] })
(backends)-[:FORK { fan: 'hosts' }]->(backend: Backend)
(backend)-[:RACE { strategy: 'health_check', timeout: 100 }]->(healthy: Healthy)
(healthy)-[:PROCESS { op: 'forward_request', input: 'request' }]->(response: Response)
(response)-[:RACE { take: 'first' }]->(fastest: FastestResponse)
(backend)-[:VENT { condition: 'unhealthy' }]->(down: BackendDown)`,
  },
  alternatives: [
    {
      language: 'Go',
      loc: 65,
      source: `// net/http, sync.RWMutex, health check goroutine, round-robin state`,
    },
    {
      language: 'Rust',
      loc: 72,
      source: `// hyper, Arc<RwLock<Vec<Backend>>>, tokio::spawn health checker`,
    },
    {
      language: 'Erlang',
      loc: 35,
      source: `// gen_server with backend list, monitor for health`,
    },
    {
      language: 'TypeScript',
      loc: 38,
      source: `// fetch + AbortController + setInterval health check`,
    },
  ],
  gnosisPrimitives: ['FORK', 'RACE', 'PROCESS', 'VENT'],
  beta1: 2,
};

// ═══════════════════════════════════════════════════════════════════════════════
// Problem 4: Circuit Breaker
// ═══════════════════════════════════════════════════════════════════════════════

const circuitBreaker: Problem = {
  name: 'Circuit Breaker',
  description:
    'Track failure rate, open circuit on threshold, half-open probe, close on success.',
  gnosis: {
    language: 'gnosis',
    loc: 8,
    source: `(request: Call)
(state: CircuitState { failures: 0, threshold: 5, window: '30s' })
(request)-[:RACE { arms: ['closed_path', 'open_path'] }]->(routed: Routed)
(routed)-[:PROCESS { op: 'call_service' }]->(result: ServiceResult)
(result)-[:FOLD { strategy: 'update_state', state: 'state' }]->(updated: StateUpdated)
(result)-[:VENT { condition: 'failure', op: 'increment_failures' }]->(failed: Failed)
(failed)-[:RACE { condition: 'over_threshold' }]->(tripped: CircuitOpen)
(tripped)-[:PROCESS { op: 'schedule_half_open', delay: '30s' }]->(probe: HalfOpen)`,
  },
  alternatives: [
    {
      language: 'Go',
      loc: 78,
      source: `// sync.Mutex, state machine, timer, atomic counters`,
    },
    {
      language: 'Rust',
      loc: 85,
      source: `// Arc<Mutex<State>>, tokio::time, state enum`,
    },
    {
      language: 'Erlang',
      loc: 45,
      source: `// gen_statem with closed/open/half_open states`,
    },
    {
      language: 'TypeScript',
      loc: 52,
      source: `// Class with state enum, setTimeout, error counting`,
    },
  ],
  gnosisPrimitives: ['RACE', 'PROCESS', 'FOLD', 'VENT'],
  beta1: 3, // three independent cycles (closed, open, half-open)
};

// ═══════════════════════════════════════════════════════════════════════════════
// Problem 5: Saga Pattern
// ═══════════════════════════════════════════════════════════════════════════════

const sagaPattern: Problem = {
  name: 'Saga Pattern',
  description: 'Multi-step transaction with compensating actions on failure.',
  gnosis: {
    language: 'gnosis',
    loc: 9,
    source: `(order: NewOrder { items: ['a','b','c'] })
(order)-[:PROCESS { op: 'reserve_inventory' }]->(reserved: Reserved)
(reserved)-[:PROCESS { op: 'charge_payment' }]->(charged: Charged)
(charged)-[:PROCESS { op: 'ship_order' }]->(shipped: Shipped)
(shipped)-[:FOLD { strategy: 'commit' }]->(done: OrderComplete)
// Compensating actions on failure (vent chain)
(shipped)-[:VENT { on: 'ship_failed' }]->(undo_charge: RefundPayment)
(charged)-[:VENT { on: 'charge_failed' }]->(undo_reserve: ReleaseInventory)
(reserved)-[:VENT { on: 'reserve_failed' }]->(cancelled: OrderCancelled)
(undo_charge)-[:PROCESS { op: 'release_inventory' }]->(cancelled)`,
  },
  alternatives: [
    {
      language: 'Go',
      loc: 92,
      source: `// Compensating functions, defer-like cleanup, state tracking`,
    },
    {
      language: 'Rust',
      loc: 88,
      source: `// trait Saga with execute/compensate, Vec<Box<dyn Saga>>`,
    },
    {
      language: 'Erlang',
      loc: 48,
      source: `// Process links + monitors for compensation`,
    },
    {
      language: 'TypeScript',
      loc: 55,
      source: `// try/catch chain with compensation stack`,
    },
  ],
  gnosisPrimitives: ['PROCESS', 'FOLD', 'VENT'],
  beta1: 3, // main path + 3 compensation branches
};

// ═══════════════════════════════════════════════════════════════════════════════
// Problem 6: CRDT Convergence (G-Counter)
// ═══════════════════════════════════════════════════════════════════════════════

const crdtConvergence: Problem = {
  name: 'CRDT Convergence',
  description:
    'G-Counter across N peers, fork state, fold to merge (max per peer).',
  gnosis: {
    language: 'gnosis',
    loc: 6,
    source: `(counter: GCounter { peers: 5 })
(counter)-[:FORK { fan: 'peers' }]->(peer: Peer)
(peer)-[:PROCESS { op: 'increment_local' }]->(updated: LocalState)
(updated)-[:FOLD { strategy: 'max_per_peer' }]->(merged: MergedCounter)
(merged)-[:PROCESS { op: 'read_total' }]->(total: Total)
(merged)-[:RACE { condition: 'all_converged' }]->(done: Converged)`,
  },
  alternatives: [
    {
      language: 'Go',
      loc: 55,
      source: `// map[PeerID]uint64, sync.RWMutex, merge function`,
    },
    {
      language: 'Rust',
      loc: 48,
      source: `// HashMap<PeerId, u64>, merge trait, CRDT library`,
    },
    {
      language: 'Erlang',
      loc: 28,
      source: `// Maps, merge function, gen_server per peer`,
    },
    {
      language: 'TypeScript',
      loc: 32,
      source: `// Map<string, number>, merge function, class wrapper`,
    },
  ],
  gnosisPrimitives: ['FORK', 'PROCESS', 'FOLD', 'RACE'],
  beta1: 1,
};

// ═══════════════════════════════════════════════════════════════════════════════
// Report
// ═══════════════════════════════════════════════════════════════════════════════

const ALL_PROBLEMS: Problem[] = [
  leaderElection,
  consensus,
  loadBalancer,
  circuitBreaker,
  sagaPattern,
  crdtConvergence,
];

function printReport(): void {
  console.log(
    '═══════════════════════════════════════════════════════════════'
  );
  console.log(' Expressiveness Benchmark');
  console.log(' gnosis .gg vs Go / Rust / Erlang / TypeScript');
  console.log(
    '═══════════════════════════════════════════════════════════════'
  );

  // LOC comparison table
  console.log('\n── Lines of Code ───────────────────────────────────────────');
  console.log(
    `${'Problem'.padEnd(20)} ${'gnosis'.padStart(8)} ${'Go'.padStart(
      8
    )} ${'Rust'.padStart(8)} ${'Erlang'.padStart(8)} ${'TS'.padStart(
      8
    )} ${'beta_1'.padStart(8)}`
  );
  console.log('─'.repeat(70));

  const gnosisTotal = { loc: 0, count: 0 };
  const langTotals: Record<string, { loc: number; count: number }> = {
    Go: { loc: 0, count: 0 },
    Rust: { loc: 0, count: 0 },
    Erlang: { loc: 0, count: 0 },
    TypeScript: { loc: 0, count: 0 },
  };

  for (const p of ALL_PROBLEMS) {
    const go = p.alternatives.find((a) => a.language === 'Go')!;
    const rust = p.alternatives.find((a) => a.language === 'Rust')!;
    const erlang = p.alternatives.find((a) => a.language === 'Erlang')!;
    const ts = p.alternatives.find((a) => a.language === 'TypeScript')!;

    console.log(
      `${p.name.padEnd(20)} ${String(p.gnosis.loc).padStart(8)} ${String(
        go.loc
      ).padStart(8)} ${String(rust.loc).padStart(8)} ${String(
        erlang.loc
      ).padStart(8)} ${String(ts.loc).padStart(8)} ${String(p.beta1).padStart(
        8
      )}`
    );

    gnosisTotal.loc += p.gnosis.loc;
    gnosisTotal.count++;
    langTotals['Go']!.loc += go.loc;
    langTotals['Go']!.count++;
    langTotals['Rust']!.loc += rust.loc;
    langTotals['Rust']!.count++;
    langTotals['Erlang']!.loc += erlang.loc;
    langTotals['Erlang']!.count++;
    langTotals['TypeScript']!.loc += ts.loc;
    langTotals['TypeScript']!.count++;
  }

  console.log('─'.repeat(70));
  console.log(
    `${'TOTAL'.padEnd(20)} ${String(gnosisTotal.loc).padStart(8)} ${String(
      langTotals['Go']!.loc
    ).padStart(8)} ${String(langTotals['Rust']!.loc).padStart(8)} ${String(
      langTotals['Erlang']!.loc
    ).padStart(8)} ${String(langTotals['TypeScript']!.loc).padStart(8)}`
  );

  // Ratio summary
  console.log(
    '\n── Expressiveness Ratios (lang/gnosis -- higher = gnosis more expressive) ──'
  );
  for (const [lang, totals] of Object.entries(langTotals)) {
    const ratio = totals.loc / gnosisTotal.loc;
    const bar = '█'.repeat(Math.round(ratio * 5));
    console.log(`  ${lang.padEnd(12)} ${ratio.toFixed(1)}x ${bar}`);
  }

  // Primitives used
  console.log(
    '\n── Gnosis Primitives Used ───────────────────────────────────'
  );
  const primitiveCounts: Record<string, number> = {};
  for (const p of ALL_PROBLEMS) {
    for (const prim of p.gnosisPrimitives) {
      primitiveCounts[prim] = (primitiveCounts[prim] ?? 0) + 1;
    }
  }
  for (const [prim, count] of Object.entries(primitiveCounts).sort(
    (a, b) => b[1] - a[1]
  )) {
    console.log(
      `  ${prim.padEnd(10)} used in ${count}/${ALL_PROBLEMS.length} problems`
    );
  }

  // Key insight
  console.log('\n── Key Insight ─────────────────────────────────────────────');
  console.log('  Every distributed systems problem above is a fork/race/fold');
  console.log('  topology. The gnosis solution formalizes the topology -- no');
  console.log('  imperative orchestration code, no manual cleanup, no');
  console.log('  AbortController, no sync primitives.');
  console.log('');
  console.log('  Structured concurrency is automatic:');
  console.log('  - RACE cancels losers (no leaked goroutines/promises)');
  console.log('  - VENT propagates failure (no unhandled rejections)');
  console.log('  - FOLD guarantees all branches complete (no dangling tasks)');
}

// ═══════════════════════════════════════════════════════════════════════════════
// Main
// ═══════════════════════════════════════════════════════════════════════════════

printReport();

export {};
