// @a0n/.ignored_gnosis/src/benchmarks/fold-training-benchmark.ts
import { readFile } from 'node:fs/promises';
import { fileURLToPath } from 'node:url';

import { parseGgProgram } from '@a0n/aeon-logic';
import { lowerUfcsSource } from '../ufcs';

import {
  bootstrapMeanConfidenceInterval,
  createDeterministicRandom,
  mean,
  shuffleInPlace,
  stdev,
  type ConfidenceInterval,
} from './statistics';

export type FoldTrainingStrategy = 'linear' | 'winner-take-all' | 'early-stop';

export interface FoldTrainingConfig {
  readonly label: 'gnosis-fold-training-benchmark-v2';
  readonly seedCount: number;
  readonly epochs: number;
  readonly learningRate: number;
  readonly exactPredictionTolerance: number;
  readonly trainAxis: readonly number[];
  readonly evalAxis: readonly number[];
  readonly samplePoints: readonly (readonly [number, number])[];
}

export interface FoldTrainingSeedMetrics {
  readonly seed: number;
  readonly trainMeanSquaredError: number;
  readonly evalMeanSquaredError: number;
  readonly exactWithinToleranceFraction: number;
  readonly cancellationLineMeanAbsoluteError: number;
}

export interface FoldTrainingSamplePrediction {
  readonly leftInput: number;
  readonly rightInput: number;
  readonly target: number;
  readonly meanPrediction: number;
  readonly predictionStddev: number;
}

export interface FoldTrainingStrategyReport {
  readonly topologyPath: string;
  readonly topologyNodeCount: number;
  readonly topologyEdgeCount: number;
  readonly structuralBeta1: number;
  readonly branchCount: number;
  readonly parameterCount: number;
  readonly seedMetrics: readonly FoldTrainingSeedMetrics[];
  readonly meanTrainMeanSquaredError: number;
  readonly stdevTrainMeanSquaredError: number;
  readonly meanEvalMeanSquaredError: number;
  readonly evalMeanSquaredErrorCi95: ConfidenceInterval;
  readonly stdevEvalMeanSquaredError: number;
  readonly meanExactWithinToleranceFraction: number;
  readonly exactWithinToleranceFractionCi95: ConfidenceInterval;
  readonly stdevExactWithinToleranceFraction: number;
  readonly meanCancellationLineMeanAbsoluteError: number;
  readonly cancellationLineMeanAbsoluteErrorCi95: ConfidenceInterval;
  readonly stdevCancellationLineMeanAbsoluteError: number;
  readonly samplePredictions: readonly FoldTrainingSamplePrediction[];
}

export interface FoldTrainingTopologyMetrics {
  readonly nodeCount: number;
  readonly edgeCount: number;
  readonly structuralBeta1: number;
  readonly branchCount: number;
  readonly parameterCount: number;
}

export interface GnosisFoldTrainingBenchmarkReport {
  readonly label: FoldTrainingConfig['label'];
  readonly config: {
    readonly seedCount: number;
    readonly epochs: number;
    readonly learningRate: number;
    readonly exactPredictionTolerance: number;
    readonly trainSampleCount: number;
    readonly evalSampleCount: number;
    readonly targetFamily: 'left-minus-right';
  };
  readonly topology: FoldTrainingTopologyMetrics;
  readonly strategies: Record<FoldTrainingStrategy, FoldTrainingStrategyReport>;
  readonly rankingByEvalMeanSquaredError: readonly FoldTrainingStrategy[];
  readonly predictedRankingMatches: boolean;
}

interface FoldTrainingSample {
  readonly leftInput: number;
  readonly rightInput: number;
  readonly target: number;
}

interface FoldTrainingParameters {
  leftWeight: number;
  leftBias: number;
  rightWeight: number;
  rightBias: number;
}

interface ParsedTopology {
  readonly strategy: FoldTrainingStrategy;
  readonly path: string;
  readonly source: string;
  readonly metrics: FoldTrainingTopologyMetrics;
}

interface PredictionResult {
  readonly leftValue: number;
  readonly rightValue: number;
  readonly prediction: number;
  readonly selectedBranch: 'left' | 'right';
}

const DEFAULT_FOLD_TRAINING_TOPOLOGY_URLS: Record<FoldTrainingStrategy, URL> = {
  linear: new URL(
    '../../examples/benchmarks/fold-training-linear.gg',
    import.meta.url
  ),
  'winner-take-all': new URL(
    '../../examples/benchmarks/fold-training-winner-take-all.gg',
    import.meta.url
  ),
  'early-stop': new URL(
    '../../examples/benchmarks/fold-training-early-stop.gg',
    import.meta.url
  ),
};

export const DEFAULT_FOLD_TRAINING_TOPOLOGY_FILES = Object.fromEntries(
  (
    Object.entries(DEFAULT_FOLD_TRAINING_TOPOLOGY_URLS) as Array<
      readonly [FoldTrainingStrategy, URL]
    >
  ).map(([strategy, url]) => [strategy, fileURLToPath(url)])
) as Record<FoldTrainingStrategy, string>;

function countConnectedComponents(
  nodeIds: readonly string[],
  edges: readonly {
    sourceIds: readonly string[];
    targetIds: readonly string[];
  }[]
): number {
  const neighbors = new Map<string, Set<string>>();
  for (const nodeId of nodeIds) {
    neighbors.set(nodeId, new Set<string>());
  }

  for (const edge of edges) {
    for (const sourceId of edge.sourceIds) {
      const sourceNeighbors = neighbors.get(sourceId);
      if (!sourceNeighbors) {
        continue;
      }
      for (const targetId of edge.targetIds) {
        sourceNeighbors.add(targetId);
        neighbors.get(targetId)?.add(sourceId);
      }
    }
  }

  let components = 0;
  const visited = new Set<string>();

  for (const nodeId of nodeIds) {
    if (visited.has(nodeId)) {
      continue;
    }

    components++;
    const frontier = [nodeId];
    while (frontier.length > 0) {
      const current = frontier.pop();
      if (current === undefined || visited.has(current)) {
        continue;
      }

      visited.add(current);
      for (const neighbor of neighbors.get(current) ?? []) {
        if (!visited.has(neighbor)) {
          frontier.push(neighbor);
        }
      }
    }
  }

  return Math.max(1, components);
}

function structuralBeta1(
  nodeIds: readonly string[],
  edges: readonly {
    sourceIds: readonly string[];
    targetIds: readonly string[];
  }[]
): number {
  const expandedEdgeCount = edges.reduce(
    (sum, edge) => sum + edge.sourceIds.length * edge.targetIds.length,
    0
  );
  return Math.max(
    0,
    expandedEdgeCount -
      nodeIds.length +
      countConnectedComponents(nodeIds, edges)
  );
}

function buildSamples(axis: readonly number[]): FoldTrainingSample[] {
  return axis.flatMap((leftInput) =>
    axis.map((rightInput) => ({
      leftInput,
      rightInput,
      target: leftInput - rightInput,
    }))
  );
}

function initializeParameters(seed: number): FoldTrainingParameters {
  const random = createDeterministicRandom(seed);
  const span = 0.8;
  const offset = -0.4;
  return {
    leftWeight: offset + random() * span,
    leftBias: offset + random() * span,
    rightWeight: offset + random() * span,
    rightBias: offset + random() * span,
  };
}

function predict(
  strategy: FoldTrainingStrategy,
  parameters: FoldTrainingParameters,
  sample: FoldTrainingSample
): PredictionResult {
  const leftValue =
    parameters.leftWeight * sample.leftInput + parameters.leftBias;
  const rightValue =
    parameters.rightWeight * sample.rightInput + parameters.rightBias;

  if (strategy === 'linear') {
    return {
      leftValue,
      rightValue,
      prediction: leftValue + rightValue,
      selectedBranch: 'left',
    };
  }

  if (strategy === 'early-stop') {
    return {
      leftValue,
      rightValue,
      prediction: leftValue,
      selectedBranch: 'left',
    };
  }

  const selectedBranch =
    Math.abs(leftValue) >= Math.abs(rightValue) ? 'left' : 'right';
  return {
    leftValue,
    rightValue,
    prediction: selectedBranch === 'left' ? leftValue : rightValue,
    selectedBranch,
  };
}

function updateParameters(
  strategy: FoldTrainingStrategy,
  parameters: FoldTrainingParameters,
  sample: FoldTrainingSample,
  learningRate: number
): void {
  const result = predict(strategy, parameters, sample);
  const error = result.prediction - sample.target;

  if (strategy === 'linear') {
    parameters.leftWeight -= learningRate * error * sample.leftInput;
    parameters.leftBias -= learningRate * error;
    parameters.rightWeight -= learningRate * error * sample.rightInput;
    parameters.rightBias -= learningRate * error;
    return;
  }

  if (result.selectedBranch === 'left') {
    parameters.leftWeight -= learningRate * error * sample.leftInput;
    parameters.leftBias -= learningRate * error;
    return;
  }

  parameters.rightWeight -= learningRate * error * sample.rightInput;
  parameters.rightBias -= learningRate * error;
}

function evaluateSamples(
  strategy: FoldTrainingStrategy,
  parameters: FoldTrainingParameters,
  samples: readonly FoldTrainingSample[],
  tolerance: number
): {
  readonly meanSquaredError: number;
  readonly exactWithinToleranceFraction: number;
  readonly cancellationLineMeanAbsoluteError: number;
} {
  if (samples.length === 0) {
    return {
      meanSquaredError: 0,
      exactWithinToleranceFraction: 0,
      cancellationLineMeanAbsoluteError: 0,
    };
  }

  let squaredErrorSum = 0;
  let exactCount = 0;
  let cancellationAbsErrorSum = 0;
  let cancellationCount = 0;

  for (const sample of samples) {
    const prediction = predict(strategy, parameters, sample).prediction;
    const error = prediction - sample.target;
    squaredErrorSum += error * error;
    if (Math.abs(error) <= tolerance) {
      exactCount++;
    }

    if (Math.abs(sample.leftInput - sample.rightInput) <= 1e-9) {
      cancellationAbsErrorSum += Math.abs(error);
      cancellationCount++;
    }
  }

  return {
    meanSquaredError: squaredErrorSum / samples.length,
    exactWithinToleranceFraction: exactCount / samples.length,
    cancellationLineMeanAbsoluteError:
      cancellationCount === 0 ? 0 : cancellationAbsErrorSum / cancellationCount,
  };
}

function summarizeSeedMetrics(
  metrics: readonly FoldTrainingSeedMetrics[]
): Omit<
  FoldTrainingStrategyReport,
  | 'topologyPath'
  | 'topologyNodeCount'
  | 'topologyEdgeCount'
  | 'structuralBeta1'
  | 'branchCount'
  | 'parameterCount'
  | 'samplePredictions'
> {
  return {
    seedMetrics: metrics,
    meanTrainMeanSquaredError: mean(
      metrics.map((entry) => entry.trainMeanSquaredError)
    ),
    stdevTrainMeanSquaredError: stdev(
      metrics.map((entry) => entry.trainMeanSquaredError)
    ),
    meanEvalMeanSquaredError: mean(
      metrics.map((entry) => entry.evalMeanSquaredError)
    ),
    evalMeanSquaredErrorCi95: bootstrapMeanConfidenceInterval(
      metrics.map((entry) => entry.evalMeanSquaredError),
      { seed: 0x11223344 }
    ),
    stdevEvalMeanSquaredError: stdev(
      metrics.map((entry) => entry.evalMeanSquaredError)
    ),
    meanExactWithinToleranceFraction: mean(
      metrics.map((entry) => entry.exactWithinToleranceFraction)
    ),
    exactWithinToleranceFractionCi95: bootstrapMeanConfidenceInterval(
      metrics.map((entry) => entry.exactWithinToleranceFraction),
      { seed: 0x22334455 }
    ),
    stdevExactWithinToleranceFraction: stdev(
      metrics.map((entry) => entry.exactWithinToleranceFraction)
    ),
    meanCancellationLineMeanAbsoluteError: mean(
      metrics.map((entry) => entry.cancellationLineMeanAbsoluteError)
    ),
    cancellationLineMeanAbsoluteErrorCi95: bootstrapMeanConfidenceInterval(
      metrics.map((entry) => entry.cancellationLineMeanAbsoluteError),
      { seed: 0x33445566 }
    ),
    stdevCancellationLineMeanAbsoluteError: stdev(
      metrics.map((entry) => entry.cancellationLineMeanAbsoluteError)
    ),
  };
}

function samplePredictions(
  strategy: FoldTrainingStrategy,
  parameterSets: readonly FoldTrainingParameters[],
  samplePoints: readonly (readonly [number, number])[]
): FoldTrainingSamplePrediction[] {
  return samplePoints.map(([leftInput, rightInput]) => {
    const sample: FoldTrainingSample = {
      leftInput,
      rightInput,
      target: leftInput - rightInput,
    };
    const predictions = parameterSets.map(
      (parameters) => predict(strategy, parameters, sample).prediction
    );
    return {
      leftInput,
      rightInput,
      target: sample.target,
      meanPrediction: mean(predictions),
      predictionStddev: stdev(predictions),
    };
  });
}

/**
 * Handles the gnosis make Default Fold Training Config workflow.
 */
export function makeDefaultFoldTrainingConfig(): FoldTrainingConfig {
  return {
    label: 'gnosis-fold-training-benchmark-v2',
    seedCount: 8,
    epochs: 720,
    learningRate: 0.035,
    exactPredictionTolerance: 0.05,
    trainAxis: [-1.5, -1.0, -0.5, 0, 0.5, 1.0, 1.5],
    evalAxis: [-1.25, -0.75, -0.25, 0.25, 0.75, 1.25],
    samplePoints: [
      [-1, -1],
      [1, -1],
      [-1, 1],
      [1, 1],
    ],
  };
}

async function loadTopology(
  strategy: FoldTrainingStrategy
): Promise<ParsedTopology> {
  const path = DEFAULT_FOLD_TRAINING_TOPOLOGY_FILES[strategy];
  const source = await readFile(path, 'utf8');
  const program = parseGgProgram(lowerUfcsSource(source));
  const foldEdges = program.edges.filter(
    (edge) => edge.type.toUpperCase() === 'FOLD'
  );
  if (foldEdges.length !== 1) {
    throw new Error(`Expected exactly one FOLD edge in ${path}`);
  }

  const foldStrategy = foldEdges[0]?.properties.strategy;
  if (foldStrategy !== strategy) {
    throw new Error(
      `Expected FOLD strategy "${strategy}" in ${path}, found "${
        foldStrategy ?? 'missing'
      }"`
    );
  }

  const branchNodes = program.nodes.filter((node) =>
    node.labels.includes('AffineBranch')
  );
  if (branchNodes.length !== 2) {
    throw new Error(`Expected exactly two AffineBranch nodes in ${path}`);
  }

  const metrics: FoldTrainingTopologyMetrics = {
    nodeCount: program.nodes.length,
    edgeCount: program.edges.length,
    structuralBeta1: structuralBeta1(
      program.nodes.map((node) => node.id),
      program.edges.map((edge) => ({
        sourceIds: edge.sourceIds,
        targetIds: edge.targetIds,
      }))
    ),
    branchCount: branchNodes.length,
    parameterCount: branchNodes.length * 2,
  };

  return {
    strategy,
    path,
    source,
    metrics,
  };
}

async function loadDefaultTopologySuite(): Promise<
  Record<FoldTrainingStrategy, ParsedTopology>
> {
  const entries = await Promise.all(
    (
      Object.keys(
        DEFAULT_FOLD_TRAINING_TOPOLOGY_FILES
      ) as FoldTrainingStrategy[]
    ).map(async (strategy) => [strategy, await loadTopology(strategy)] as const)
  );

  return Object.fromEntries(entries) as Record<
    FoldTrainingStrategy,
    ParsedTopology
  >;
}

function assertSharedTopology(
  topologies: Record<FoldTrainingStrategy, ParsedTopology>
): FoldTrainingTopologyMetrics {
  const baseline = topologies.linear.metrics;
  for (const strategy of Object.keys(topologies) as FoldTrainingStrategy[]) {
    const metrics = topologies[strategy].metrics;
    const isShared =
      metrics.nodeCount === baseline.nodeCount &&
      metrics.edgeCount === baseline.edgeCount &&
      metrics.structuralBeta1 === baseline.structuralBeta1 &&
      metrics.branchCount === baseline.branchCount &&
      metrics.parameterCount === baseline.parameterCount;
    if (!isShared) {
      throw new Error(
        `Topology mismatch for ${strategy}; benchmark is not parameter-matched`
      );
    }
  }
  return baseline;
}

function trainStrategy(
  strategy: FoldTrainingStrategy,
  config: FoldTrainingConfig,
  trainSamples: readonly FoldTrainingSample[],
  evalSamples: readonly FoldTrainingSample[]
): {
  readonly seedMetrics: readonly FoldTrainingSeedMetrics[];
  readonly trainedParameters: readonly FoldTrainingParameters[];
} {
  const seedMetrics: FoldTrainingSeedMetrics[] = [];
  const trainedParameters: FoldTrainingParameters[] = [];

  for (let seedIndex = 0; seedIndex < config.seedCount; seedIndex++) {
    const seed = seedIndex + 1;
    const parameters = initializeParameters(seed);
    const random = createDeterministicRandom(seed * 0x9e3779b1);

    for (let epoch = 0; epoch < config.epochs; epoch++) {
      const orderedSamples = [...trainSamples];
      shuffleInPlace(orderedSamples, random);
      for (const sample of orderedSamples) {
        updateParameters(strategy, parameters, sample, config.learningRate);
      }
    }

    const trainEvaluation = evaluateSamples(
      strategy,
      parameters,
      trainSamples,
      config.exactPredictionTolerance
    );
    const evalEvaluation = evaluateSamples(
      strategy,
      parameters,
      evalSamples,
      config.exactPredictionTolerance
    );

    trainedParameters.push({ ...parameters });
    seedMetrics.push({
      seed,
      trainMeanSquaredError: trainEvaluation.meanSquaredError,
      evalMeanSquaredError: evalEvaluation.meanSquaredError,
      exactWithinToleranceFraction: evalEvaluation.exactWithinToleranceFraction,
      cancellationLineMeanAbsoluteError:
        evalEvaluation.cancellationLineMeanAbsoluteError,
    });
  }

  return {
    seedMetrics,
    trainedParameters,
  };
}

/**
 * Handles the gnosis run Gnosis Fold Training Benchmark workflow.
 */
export async function runGnosisFoldTrainingBenchmark(
  config: FoldTrainingConfig = makeDefaultFoldTrainingConfig()
): Promise<GnosisFoldTrainingBenchmarkReport> {
  const topologies = await loadDefaultTopologySuite();
  const sharedTopology = assertSharedTopology(topologies);
  const trainSamples = buildSamples(config.trainAxis);
  const evalSamples = buildSamples(config.evalAxis);

  const strategies = {} as Record<
    FoldTrainingStrategy,
    FoldTrainingStrategyReport
  >;
  for (const strategy of Object.keys(topologies) as FoldTrainingStrategy[]) {
    const trained = trainStrategy(strategy, config, trainSamples, evalSamples);
    const summary = summarizeSeedMetrics(trained.seedMetrics);
    strategies[strategy] = {
      topologyPath: topologies[strategy].path,
      topologyNodeCount: topologies[strategy].metrics.nodeCount,
      topologyEdgeCount: topologies[strategy].metrics.edgeCount,
      structuralBeta1: topologies[strategy].metrics.structuralBeta1,
      branchCount: topologies[strategy].metrics.branchCount,
      parameterCount: topologies[strategy].metrics.parameterCount,
      ...summary,
      samplePredictions: samplePredictions(
        strategy,
        trained.trainedParameters,
        config.samplePoints
      ),
    };
  }

  const rankingByEvalMeanSquaredError = (
    Object.keys(strategies) as FoldTrainingStrategy[]
  ).sort(
    (left, right) =>
      strategies[left].meanEvalMeanSquaredError -
      strategies[right].meanEvalMeanSquaredError
  );

  return {
    label: config.label,
    config: {
      seedCount: config.seedCount,
      epochs: config.epochs,
      learningRate: config.learningRate,
      exactPredictionTolerance: config.exactPredictionTolerance,
      trainSampleCount: trainSamples.length,
      evalSampleCount: evalSamples.length,
      targetFamily: 'left-minus-right',
    },
    topology: sharedTopology,
    strategies,
    rankingByEvalMeanSquaredError,
    predictedRankingMatches:
      rankingByEvalMeanSquaredError[0] === 'linear' &&
      rankingByEvalMeanSquaredError[1] === 'winner-take-all' &&
      rankingByEvalMeanSquaredError[2] === 'early-stop',
  };
}

function formatSamplePair(sample: FoldTrainingSamplePrediction): string {
  return `(${sample.leftInput.toFixed(2)}, ${sample.rightInput.toFixed(2)})`;
}

/**
 * Handles the gnosis render Gnosis Fold Training Benchmark Markdown workflow.
 */
export function renderGnosisFoldTrainingBenchmarkMarkdown(
  report: GnosisFoldTrainingBenchmarkReport
): string {
  const lines: string[] = [];
  lines.push('# Gnosis Fold Training Benchmark');
  lines.push('');
  lines.push(`- Label: \`${report.label}\``);
  lines.push(`- Target family: \`${report.config.targetFamily}\``);
  lines.push(`- Seeds: \`${report.config.seedCount}\``);
  lines.push(`- Epochs: \`${report.config.epochs}\``);
  lines.push(`- Learning rate: \`${report.config.learningRate}\``);
  lines.push(`- Train samples: \`${report.config.trainSampleCount}\``);
  lines.push(`- Eval samples: \`${report.config.evalSampleCount}\``);
  lines.push(`- Shared parameter count: \`${report.topology.parameterCount}\``);
  lines.push(`- Shared structural β₁: \`${report.topology.structuralBeta1}\``);
  lines.push(
    `- Predicted ranking recovered: \`${
      report.predictedRankingMatches ? 'yes' : 'no'
    }\``
  );
  lines.push('');
  lines.push('## Aggregated Metrics');
  lines.push('');
  lines.push(
    '| Strategy | Mean eval MSE | Eval MSE 95% CI | Eval MSE σ | Exact-within-tolerance | Exact 95% CI | Cancellation-line abs error | Cancellation 95% CI |'
  );
  lines.push('|---|---:|---:|---:|---:|---:|---:|---:|');

  for (const strategy of report.rankingByEvalMeanSquaredError) {
    const metrics = report.strategies[strategy];
    lines.push(
      `| \`${strategy}\` | ${metrics.meanEvalMeanSquaredError.toFixed(
        3
      )} | [${metrics.evalMeanSquaredErrorCi95.low.toFixed(
        3
      )}, ${metrics.evalMeanSquaredErrorCi95.high.toFixed(
        3
      )}] | ${metrics.stdevEvalMeanSquaredError.toFixed(
        3
      )} | ${metrics.meanExactWithinToleranceFraction.toFixed(
        3
      )} | [${metrics.exactWithinToleranceFractionCi95.low.toFixed(
        3
      )}, ${metrics.exactWithinToleranceFractionCi95.high.toFixed(
        3
      )}] | ${metrics.meanCancellationLineMeanAbsoluteError.toFixed(
        3
      )} | [${metrics.cancellationLineMeanAbsoluteErrorCi95.low.toFixed(
        3
      )}, ${metrics.cancellationLineMeanAbsoluteErrorCi95.high.toFixed(3)}] |`
    );
  }

  lines.push('');
  lines.push('## Sample Predictions');
  lines.push('');
  lines.push(
    '| Strategy | Input pair | Target | Mean prediction | Prediction σ |'
  );
  lines.push('|---|---|---:|---:|---:|');

  for (const strategy of report.rankingByEvalMeanSquaredError) {
    for (const sample of report.strategies[strategy].samplePredictions) {
      lines.push(
        `| \`${strategy}\` | ${formatSamplePair(
          sample
        )} | ${sample.target.toFixed(3)} | ${sample.meanPrediction.toFixed(
          3
        )} | ${sample.predictionStddev.toFixed(3)} |`
      );
    }
  }

  lines.push('');
  lines.push(
    'Interpretation: the topology, parameter count, and data are held fixed across three Gnosis programs; only the fold strategy changes. Linear recombination learns the cancellation-sensitive target family, while nonlinear selection leaves a persistent recombination error floor.'
  );
  lines.push('');
  return `${lines.join('\n')}\n`;
}
