// @a0n/.ignored_gnosis/src/benchmarks/hetero-moa-fabric-benchmark.ts
import {
  HeteroMoAFabricCommunityMemory,
  runHeteroMoAFabric,
  type HeteroMoAFabricBackendKind,
  type HeteroMoAFabricBackendScore,
  type HeteroMoAFabricBackendDescriptor,
  type HeteroMoAFabricLaneCounts,
  type HeteroMoAFabricLaunchScheduleEntry,
  type HeteroMoAFabricLayerKind,
} from '../runtime/hetero-fabric';
import { init } from '../neural-compat';

export interface HeteroMoAFabricBenchmarkConfig {
  readonly iterations: number;
  readonly hedgeDelayMs: number;
  readonly laneCounts: {
    cpu: number;
    gpu: number;
    npu: number;
    wasm: number;
  };
  readonly layerLatencyMs: Record<HeteroMoAFabricLayerKind, number>;
}

export interface HeteroMoAFabricLiveBenchmarkConfig
  extends Partial<HeteroMoAFabricBenchmarkConfig> {
  readonly inputLength?: number;
  readonly includeEnvCommandBackends?: boolean;
  readonly backends?: HeteroMoAFabricBackendDescriptor[];
  readonly fabricKey?: string;
  readonly engineFactory?: () => Promise<HeteroMoAFabricLiveBenchmarkEngine>;
}

export interface HeteroMoAFabricCloudRunStamp {
  readonly detected: boolean;
  readonly service?: string;
  readonly revision?: string;
  readonly job?: string;
  readonly taskIndex?: string;
  readonly region?: string;
  readonly project?: string;
}

export interface HeteroMoAFabricBenchmarkBackendSummary {
  readonly id: string;
  readonly label: string;
  readonly kind: HeteroMoAFabricBackendKind;
  readonly layer: HeteroMoAFabricLayerKind;
  readonly priority?: number;
  readonly detail?: string;
  readonly predictedLatencyMs?: number;
}

export interface HeteroMoAFabricBenchmarkDiscovery {
  readonly backendCount: number;
  readonly availableLaneCounts: HeteroMoAFabricLaneCounts;
  readonly activeLaneCounts: HeteroMoAFabricLaneCounts;
  readonly backends: readonly HeteroMoAFabricBenchmarkBackendSummary[];
}

export interface HeteroMoAFabricBenchmarkReport {
  readonly label: 'gnosis-hetero-moa-fabric-benchmark-v1';
  readonly mode: 'synthetic' | 'live';
  readonly iterations: number;
  readonly inputLength: number;
  readonly cloudRun: HeteroMoAFabricCloudRunStamp;
  readonly discovery: HeteroMoAFabricBenchmarkDiscovery;
  readonly baseline: {
    readonly meanWallTimeMs: number;
  };
  readonly fabric: {
    readonly meanWallTimeMs: number;
    readonly meanCpuTimeMs: number;
    readonly meanSkippedHedges: number;
    readonly meanVentShare: number;
    readonly meanQueueOccupancy: number;
    readonly winnerCounts: Record<string, number>;
  };
  readonly community: {
    readonly finalScores: Record<string, HeteroMoAFabricBackendScore>;
    readonly finalLaunchSchedule: readonly HeteroMoAFabricLaunchScheduleEntry[];
  };
  readonly relativeSpeedup: number;
}

interface HeteroMoAFabricLiveBenchmarkEngine {
  getHeteroFabricBackends(options?: {
    includeEnvCommandBackends?: boolean;
    backends?: HeteroMoAFabricBackendDescriptor[];
  }): Promise<HeteroMoAFabricBackendDescriptor[]>;
}

interface HeteroMoAFabricBenchmarkRunRequest {
  readonly mode: 'synthetic' | 'live';
  readonly iterations: number;
  readonly inputLength: number;
  readonly hedgeDelayMs: number;
  readonly fabricKey: string;
  readonly backends: readonly HeteroMoAFabricBackendDescriptor[];
  readonly availableLaneCounts: HeteroMoAFabricLaneCounts;
  readonly activeLaneCounts: HeteroMoAFabricLaneCounts;
  readonly env: Record<string, string | undefined>;
}

const DEFAULT_CONFIG: HeteroMoAFabricBenchmarkConfig = {
  iterations: 8,
  hedgeDelayMs: 1,
  laneCounts: {
    cpu: 1,
    gpu: 1,
    npu: 1,
    wasm: 1,
  },
  layerLatencyMs: {
    cpu: 4,
    gpu: 2,
    npu: 3,
    wasm: 5,
  },
};
const DEFAULT_LIVE_INPUT_LENGTH = 224;

function nowMs(): number {
  return globalThis.performance?.now?.() ?? Date.now();
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => {
    setTimeout(resolve, ms);
  });
}

function mean(values: readonly number[]): number {
  if (values.length === 0) {
    return 0;
  }
  return values.reduce((sum, value) => sum + value, 0) / values.length;
}

function createBenchmarkInput(
  iteration: number,
  inputLength: number
): Float32Array {
  const input = new Float32Array(inputLength);
  for (let index = 0; index < inputLength; index++) {
    const phase = ((iteration + 1) * (index + 3)) % 17;
    input[index] = (phase - 8) / 8;
  }
  return input;
}

function summarizeBackends(
  backends: readonly HeteroMoAFabricBackendDescriptor[]
): readonly HeteroMoAFabricBenchmarkBackendSummary[] {
  return backends.map((backend) => ({
    id: backend.id,
    label: backend.label,
    kind: backend.kind,
    layer: backend.layer,
    priority: backend.priority,
    detail: backend.detail,
    predictedLatencyMs: backend.predictedLatencyMs,
  }));
}

function countLaneBackends(
  backends: readonly Pick<HeteroMoAFabricBackendDescriptor, 'layer'>[]
): HeteroMoAFabricLaneCounts {
  return {
    cpu: backends.filter((backend) => backend.layer === 'cpu').length,
    gpu: backends.filter((backend) => backend.layer === 'gpu').length,
    npu: backends.filter((backend) => backend.layer === 'npu').length,
    wasm: backends.filter((backend) => backend.layer === 'wasm').length,
  };
}

function clampLaneCounts(
  requested: Partial<HeteroMoAFabricLaneCounts> | undefined,
  available: HeteroMoAFabricLaneCounts
): HeteroMoAFabricLaneCounts {
  return {
    cpu: Math.max(0, Math.min(requested?.cpu ?? available.cpu, available.cpu)),
    gpu: Math.max(0, Math.min(requested?.gpu ?? available.gpu, available.gpu)),
    npu: Math.max(0, Math.min(requested?.npu ?? available.npu, available.npu)),
    wasm: Math.max(
      0,
      Math.min(requested?.wasm ?? available.wasm, available.wasm)
    ),
  };
}

/**
 * Handles the gnosis detect Cloud Run Stamp workflow.
 */
export function detectCloudRunStamp(
  env: Record<string, string | undefined> = process.env
): HeteroMoAFabricCloudRunStamp {
  const detected = Boolean(env.K_SERVICE || env.CLOUD_RUN_JOB);
  return {
    detected,
    service: env.K_SERVICE,
    revision: env.K_REVISION,
    job: env.CLOUD_RUN_JOB,
    taskIndex: env.CLOUD_RUN_TASK_INDEX,
    region: env.CLOUD_RUN_REGION ?? env.GOOGLE_CLOUD_REGION,
    project: env.GOOGLE_CLOUD_PROJECT,
  };
}

function createSyntheticBackends(
  config: HeteroMoAFabricBenchmarkConfig
): HeteroMoAFabricBackendDescriptor[] {
  const descriptors: HeteroMoAFabricBackendDescriptor[] = [];
  const layers: readonly HeteroMoAFabricLayerKind[] = [
    'cpu',
    'gpu',
    'npu',
    'wasm',
  ];

  for (const layer of layers) {
    const kind =
      layer === 'cpu'
        ? 'cpu'
        : layer === 'gpu'
        ? 'webgpu'
        : layer === 'npu'
        ? 'webnn'
        : 'wasm';
    descriptors.push({
      id: `${layer}-synthetic`,
      label: `${layer.toUpperCase()} synthetic lane`,
      kind,
      layer,
      predictedLatencyMs: config.layerLatencyMs[layer],
      priority: layer === 'gpu' ? 3 : layer === 'npu' ? 2 : 1,
      async execute(input) {
        await sleep(config.layerLatencyMs[layer]);
        return {
          value: {
            layer,
            input,
          },
          cpuTimeMs: config.layerLatencyMs[layer],
        };
      },
    });
  }

  return descriptors;
}

async function runBaselineSequential(
  backends: readonly HeteroMoAFabricBackendDescriptor[],
  input: Float32Array
): Promise<number> {
  const startedAt = nowMs();
  for (const backend of backends) {
    await backend.execute(input, {
      signal: new AbortController().signal,
      role: 'primary',
      lane: 0,
      layer: backend.layer,
      cursorPosition: 0,
      frameProtocol: 'aeon-10-byte-binary',
    });
  }
  return Math.max(0, nowMs() - startedAt);
}

async function runBenchmarkReport(
  request: HeteroMoAFabricBenchmarkRunRequest
): Promise<HeteroMoAFabricBenchmarkReport> {
  const {
    mode,
    iterations,
    inputLength,
    hedgeDelayMs,
    fabricKey,
    backends,
    availableLaneCounts,
    activeLaneCounts,
    env,
  } = request;
  const communityMemory = new HeteroMoAFabricCommunityMemory({
    decayHalfLifeMs: 60_000,
  });
  const baselineWallTimes: number[] = [];
  const fabricWallTimes: number[] = [];
  const fabricCpuTimes: number[] = [];
  const skippedHedges: number[] = [];
  const ventShares: number[] = [];
  const queueOccupancies: number[] = [];
  const winnerCounts = new Map<string, number>();

  for (let iteration = 0; iteration < iterations; iteration++) {
    const input = createBenchmarkInput(iteration, inputLength);
    baselineWallTimes.push(await runBaselineSequential(backends, input));
    const fabricStartedAt = nowMs();
    const result = await runHeteroMoAFabric(backends, input, {
      fabricKey,
      communityMemory,
      plan: {
        hedgeDelayMs,
        laneCounts: activeLaneCounts,
      },
    });
    fabricWallTimes.push(Math.max(0, nowMs() - fabricStartedAt));
    fabricCpuTimes.push(result.telemetry.cpuTimeMs);
    skippedHedges.push(result.telemetry.skippedHedges);
    ventShares.push(result.telemetry.ventShare);
    queueOccupancies.push(result.telemetry.queueOccupancy);
    if (result.winner) {
      winnerCounts.set(
        result.winner.layer,
        (winnerCounts.get(result.winner.layer) ?? 0) + 1
      );
    }
  }

  const baselineMean = mean(baselineWallTimes);
  const fabricMean = mean(fabricWallTimes);
  const finalScores = communityMemory.snapshotScores(fabricKey);
  const finalLaunchSchedule = communityMemory.createLaunchSchedule(
    fabricKey,
    backends
  );
  return {
    label: 'gnosis-hetero-moa-fabric-benchmark-v1',
    mode,
    iterations,
    inputLength,
    cloudRun: detectCloudRunStamp(env),
    discovery: {
      backendCount: backends.length,
      availableLaneCounts,
      activeLaneCounts,
      backends: summarizeBackends(backends),
    },
    baseline: {
      meanWallTimeMs: baselineMean,
    },
    fabric: {
      meanWallTimeMs: fabricMean,
      meanCpuTimeMs: mean(fabricCpuTimes),
      meanSkippedHedges: mean(skippedHedges),
      meanVentShare: mean(ventShares),
      meanQueueOccupancy: mean(queueOccupancies),
      winnerCounts: Object.fromEntries(winnerCounts),
    },
    community: {
      finalScores,
      finalLaunchSchedule,
    },
    relativeSpeedup: fabricMean > 0 ? baselineMean / fabricMean : 0,
  };
}

/**
 * Handles the gnosis run Hetero Mo AFabric Benchmark workflow.
 */
export async function runHeteroMoAFabricBenchmark(
  overrides: Partial<HeteroMoAFabricBenchmarkConfig> = {},
  env: Record<string, string | undefined> = process.env
): Promise<HeteroMoAFabricBenchmarkReport> {
  const config: HeteroMoAFabricBenchmarkConfig = {
    ...DEFAULT_CONFIG,
    ...overrides,
    laneCounts: {
      ...DEFAULT_CONFIG.laneCounts,
      ...overrides.laneCounts,
    },
    layerLatencyMs: {
      ...DEFAULT_CONFIG.layerLatencyMs,
      ...overrides.layerLatencyMs,
    },
  };
  const backends = createSyntheticBackends(config);
  const availableLaneCounts = countLaneBackends(backends);
  const activeLaneCounts = clampLaneCounts(
    config.laneCounts,
    availableLaneCounts
  );
  return runBenchmarkReport({
    mode: 'synthetic',
    iterations: config.iterations,
    inputLength: 3,
    hedgeDelayMs: config.hedgeDelayMs,
    fabricKey: 'benchmark-fabric',
    backends,
    availableLaneCounts,
    activeLaneCounts,
    env,
  });
}

/**
 * Handles the gnosis run Live Hetero Mo AFabric Benchmark workflow.
 */
export async function runLiveHeteroMoAFabricBenchmark(
  overrides: HeteroMoAFabricLiveBenchmarkConfig = {},
  env: Record<string, string | undefined> = process.env
): Promise<HeteroMoAFabricBenchmarkReport> {
  const engine = await (overrides.engineFactory ?? init)();
  const backends = (
    await engine.getHeteroFabricBackends({
      includeEnvCommandBackends: overrides.includeEnvCommandBackends,
      backends: overrides.backends,
    })
  ).filter((backend) => backend.available !== false);
  const availableLaneCounts = countLaneBackends(backends);
  const activeLaneCounts = clampLaneCounts(
    overrides.laneCounts,
    availableLaneCounts
  );

  return runBenchmarkReport({
    mode: 'live',
    iterations: Math.max(
      1,
      Math.floor(overrides.iterations ?? DEFAULT_CONFIG.iterations)
    ),
    inputLength: Math.max(
      1,
      Math.floor(overrides.inputLength ?? DEFAULT_LIVE_INPUT_LENGTH)
    ),
    hedgeDelayMs: Math.max(
      0,
      overrides.hedgeDelayMs ?? DEFAULT_CONFIG.hedgeDelayMs
    ),
    fabricKey: overrides.fabricKey?.trim() || 'live-benchmark-fabric',
    backends,
    availableLaneCounts,
    activeLaneCounts,
    env,
  });
}

const isDirectExecution =
  typeof process !== 'undefined' &&
  Array.isArray(process.argv) &&
  typeof process.argv[1] === 'string' &&
  import.meta.url === new URL(process.argv[1], 'file:').href;

if (isDirectExecution) {
  const argv = process.argv.slice(2);
  const report = argv.includes('--live')
    ? await runLiveHeteroMoAFabricBenchmark()
    : await runHeteroMoAFabricBenchmark();
  console.log(JSON.stringify(report, null, 2));
}
