// @a0n/.ignored_gnosis/src/benchmarks/native-runtime-benchmark.ts
#!/usr/bin/env node
/**
 * Native Runtime Benchmark
 *
 * Measures the accuracy of gnosis-polyglot's static topology extraction
 * against empirical observations produced by running the same source files
 * through their native runtimes (Python 3, rustc + run, go run).
 *
 * Methodology (vibe-science):
 *   1. OBSERVE  — run the fixture source through the native runtime,
 *                 capture stdout/stderr, exit code, and wall-clock time.
 *   2. EXTRACT  — run the same fixture through gnosis-polyglot to get
 *                 a PolyglotAnalysisResult with per-function topologies.
 *   3. PREDICT  — from the topology, derive predicted runtime properties:
 *                 function count, resource acquire/release pairs, async
 *                 functions, branches, return paths, ghost-mass balance.
 *   4. COMPARE  — score each prediction against the observed empirical fact.
 *   5. REPORT   — emit pass/fail matrix + aggregate precision score.
 *
 * Why this matters:
 *   A topology that doesn't predict what the native runtime does is useless
 *   for formal verification. This benchmark gives a falsifiable scorecard.
 *
 * Usage (gnode):
 *   node ./bin/gnode.js run src/benchmarks/native-runtime-benchmark.ts
 *
 * Usage (tsx):
 *   npx tsx src/benchmarks/native-runtime-benchmark.ts
 */

import { execSync } from 'child_process';
import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import { analyzePolyglotSourceString } from '../polyglot-bridge';

// ── Types ──────────────────────────────────────────────────────────────────

export interface RuntimeObservation {
  language: string;
  exitCode: number;
  stdout: string;
  stderr: string;
  wallClockMs: number;
  /** Key-value facts emitted to stdout by the fixture (KEY:VALUE lines). */
  facts: Record<string, string | number | boolean>;
}

export interface TopologyPrediction {
  language: string;
  functionCount: number;
  asyncFunctions: number;
  /** Count of nodes with 'Acquire' label (context managers, lock acquires…) */
  acquisitions: number;
  /** Count of nodes with 'Release' label */
  releases: number;
  branchCount: number;
  returnPaths: number;
  /** Total ghost mass debt across all functions (resource leak signal) */
  totalLeakDebt: number;
  diagnostics: string[];
}

export interface PredictionMatch {
  property: string;
  predicted: string | number | boolean;
  observed: string | number | boolean;
  pass: boolean;
}

export interface NativeRuntimeResult {
  name: string;
  language: string;
  observation: RuntimeObservation;
  prediction: TopologyPrediction;
  matches: PredictionMatch[];
  precisionScore: number; // 0.0–1.0
  observationMs: number;
  extractionMs: number;
}

// ── Fixture source files ────────────────────────────────────────────────────

/**
 * Python fixture: context-manager resources + threading fan-out.
 * Each observable fact is printed as KEY:VALUE so we can parse it.
 */
const PYTHON_FIXTURE = `#!/usr/bin/env python3
"""gnosis_native_bm_py — emits observable facts for the native runtime bench."""
import threading

ACQUIRED = []
RELEASED = []

class Resource:
    def __init__(self, name):
        self.name = name

    def __enter__(self):
        ACQUIRED.append(self.name)
        return self

    def __exit__(self, *_):
        RELEASED.append(self.name)
        return False

def fetch(resource_name: str) -> str:
    with Resource(resource_name) as r:
        return f"ok:{r.name}"

def process_data(data: list) -> dict:
    if not data:
        raise ValueError("empty")
    result = {}
    for item in data:
        result[item] = fetch(item)
    return result

def parallel_fetch(items: list) -> list:
    results = []
    threads = []
    def worker(name):
        results.append(fetch(name))
    for item in items:
        t = threading.Thread(target=worker, args=(item,))
        threads.append(t)
        t.start()
    for t in threads:
        t.join()
    return results

if __name__ == '__main__':
    r1 = process_data(["alpha", "beta"])
    r2 = parallel_fetch(["gamma", "delta"])
    print(f"FUNCTION_COUNT:4")
    print(f"ACQUIRED:{len(ACQUIRED)}")
    print(f"RELEASED:{len(RELEASED)}")
    print(f"BALANCED:{str(len(ACQUIRED) == len(RELEASED)).lower()}")
    print(f"THREADS_SPAWNED:2")
    print(f"RESULTS:{len(r1) + len(r2)}")
`;

/**
 * Rust fixture: error path + sequential map (no tokio required for fixture).
 */
const RUST_FIXTURE = `use std::collections::HashMap;

fn fetch(name: &str) -> Result<String, String> {
    Ok(format!("ok:{}", name))
}

fn process_data(data: &[&str]) -> Result<HashMap<String, String>, String> {
    if data.is_empty() {
        return Err("empty".to_string());
    }
    let mut result = HashMap::new();
    for item in data {
        result.insert(item.to_string(), fetch(item)?);
    }
    Ok(result)
}

fn parallel_fetch_sync(items: &[&str]) -> Vec<String> {
    items.iter().map(|name| fetch(name).unwrap_or_default()).collect()
}

fn main() {
    let r1 = process_data(&["alpha", "beta"]).expect("process failed");
    let r2 = parallel_fetch_sync(&["gamma", "delta"]);
    println!("FUNCTION_COUNT:3");
    println!("RESULTS:{}", r1.len() + r2.len());
    println!("ERRORS:0");
}
`;

/**
 * Go fixture: goroutine fan-out with WaitGroup (idiomatic parallel pattern).
 */
const GO_FIXTURE = `package main

import (
\t"fmt"
\t"sync"
)

func fetch(name string) string {
\treturn "ok:" + name
}

func processData(data []string) map[string]string {
\tif len(data) == 0 {
\t\tpanic("empty")
\t}
\tresult := make(map[string]string)
\tfor _, item := range data {
\t\tresult[item] = fetch(item)
\t}
\treturn result
}

func parallelFetch(items []string) []string {
\tresults := make([]string, len(items))
\tvar wg sync.WaitGroup
\tfor i, item := range items {
\t\twg.Add(1)
\t\tgo func(idx int, name string) {
\t\t\tdefer wg.Done()
\t\t\tresults[idx] = fetch(name)
\t\t}(i, item)
\t}
\twg.Wait()
\treturn results
}

func main() {
\tr1 := processData([]string{"alpha", "beta"})
\tr2 := parallelFetch([]string{"gamma", "delta"})
\tfmt.Printf("FUNCTION_COUNT:4\\n")
\tfmt.Printf("RESULTS:%d\\n", len(r1)+len(r2))
\tfmt.Printf("GOROUTINES_SPAWNED:2\\n")
\tfmt.Printf("SYNC_JOINS:1\\n")
}
`;

// ── Native runtime execution ──────────────────────────────────────────────

function parseObservationFacts(
  stdout: string
): Record<string, string | number | boolean> {
  const facts: Record<string, string | number | boolean> = {};
  for (const line of stdout.split('\n')) {
    const colonIdx = line.indexOf(':');
    if (colonIdx < 0) continue;
    const key = line.slice(0, colonIdx).trim();
    const val = line.slice(colonIdx + 1).trim();
    if (!key) continue;
    const num = Number(val);
    if (!isNaN(num) && val !== '') facts[key] = num;
    else if (val === 'true') facts[key] = true;
    else if (val === 'false') facts[key] = false;
    else facts[key] = val;
  }
  return facts;
}

function runNative(
  lang: string,
  source: string,
  ext: string
): RuntimeObservation {
  const tmpDir = fs.mkdtempSync(
    path.join(os.tmpdir(), `gnosis-native-bm-${lang}-`)
  );
  const srcFile = path.join(tmpDir, `fixture.${ext}`);
  fs.writeFileSync(srcFile, source, 'utf8');

  let cmd: string;
  if (lang === 'rust') {
    const binary = path.join(tmpDir, 'fixture_bin');
    try {
      execSync(`rustc "${srcFile}" -o "${binary}"`, {
        timeout: 60_000,
        stdio: 'pipe',
      });
    } catch (e: unknown) {
      const ce = e as { stdout?: Buffer; stderr?: Buffer };
      try {
        fs.rmSync(tmpDir, { recursive: true, force: true });
      } catch {}
      return {
        language: lang,
        exitCode: 1,
        stdout: '',
        stderr: ce.stderr?.toString() ?? 'compile failed',
        wallClockMs: 0,
        facts: {},
      };
    }
    cmd = `"${binary}"`;
  } else if (lang === 'python') {
    cmd = `python3 "${srcFile}"`;
  } else if (lang === 'go') {
    cmd = `go run "${srcFile}"`;
  } else {
    throw new Error(`Unsupported language: ${lang}`);
  }

  const t0 = performance.now();
  let stdout = '';
  let stderr = '';
  let exitCode = 0;
  try {
    stdout = execSync(cmd, { timeout: 30_000, stdio: 'pipe' }).toString();
  } catch (e: unknown) {
    const ce = e as { stdout?: Buffer; stderr?: Buffer; status?: number };
    stdout = ce.stdout?.toString() ?? '';
    stderr = ce.stderr?.toString() ?? '';
    exitCode = ce.status ?? 1;
  }
  const wallClockMs = performance.now() - t0;
  try {
    fs.rmSync(tmpDir, { recursive: true, force: true });
  } catch {}

  return {
    language: lang,
    exitCode,
    stdout,
    stderr,
    wallClockMs,
    facts: parseObservationFacts(stdout),
  };
}

// ── Topology extraction ──────────────────────────────────────────────────

async function extractTopology(
  lang: string,
  source: string,
  fileName: string
): Promise<{ prediction: TopologyPrediction; extractionMs: number }> {
  const t0 = performance.now();
  let functionCount = 0;
  let asyncFunctions = 0;
  let acquisitions = 0;
  let releases = 0;
  let branchCount = 0;
  let returnPaths = 0;
  let totalLeakDebt = 0;
  const diagnostics: string[] = [];

  try {
    const result = await analyzePolyglotSourceString(source, fileName);

    for (const err of result.errors) {
      diagnostics.push(err);
    }

    functionCount = result.functions.length;

    for (const fn of result.functions) {
      if (fn.signature?.is_async) asyncFunctions++;
      totalLeakDebt += fn.ghostMassLedger?.leak_debt ?? 0;
      for (const diag of fn.diagnostics) {
        if (diag.severity === 'Error') diagnostics.push(diag.message);
      }
      // Walk topology nodes from the GraphAST
      for (const [_id, node] of fn.ast.nodes) {
        const labels = node.labels ?? [];
        if (labels.includes('Acquire')) acquisitions++;
        if (labels.includes('Release')) releases++;
        if (labels.includes('Branch') || labels.includes('Condition'))
          branchCount++;
        if (labels.includes('Return')) returnPaths++;
      }
    }
  } catch (e: unknown) {
    diagnostics.push(String(e));
  }

  const extractionMs = performance.now() - t0;
  return {
    prediction: {
      language: lang,
      functionCount,
      asyncFunctions,
      acquisitions,
      releases,
      branchCount,
      returnPaths,
      totalLeakDebt,
      diagnostics,
    },
    extractionMs,
  };
}

// ── Prediction matching ───────────────────────────────────────────────────

function matchPredictions(
  lang: string,
  prediction: TopologyPrediction,
  observation: RuntimeObservation
): PredictionMatch[] {
  const matches: PredictionMatch[] = [];

  const check = (
    property: string,
    predicted: string | number | boolean,
    observed: string | number | boolean,
    pass: boolean
  ) => matches.push({ property, predicted, observed, pass });

  // ── Universal checks ──
  // Topology is non-empty
  check(
    'topology_non_empty',
    prediction.functionCount,
    '>= 1',
    prediction.functionCount >= 1
  );

  // Native runtime exits successfully
  check('native_exits_0', observation.exitCode, 0, observation.exitCode === 0);

  // Function count matches fixture-reported count
  const reportedFnCount = observation.facts['FUNCTION_COUNT'] as
    | number
    | undefined;
  if (reportedFnCount !== undefined) {
    check(
      'function_count',
      prediction.functionCount,
      reportedFnCount,
      prediction.functionCount === reportedFnCount
    );
  }

  // Total results matched (fixture prints RESULTS:N)
  const reportedResults = observation.facts['RESULTS'] as number | undefined;
  if (reportedResults !== undefined) {
    // Topology should have at least 1 return path per function-that-returns
    check(
      'return_paths_gte_1',
      prediction.returnPaths,
      '>= 1',
      prediction.returnPaths >= 1
    );
    check('results_non_zero', reportedResults, '>= 1', reportedResults >= 1);
  }

  // ── Python-specific ──
  if (lang === 'python') {
    const balanced = observation.facts['BALANCED'];
    check(
      'resources_balanced',
      prediction.acquisitions === prediction.releases,
      balanced === true,
      // Topology predicts balanced iff acquire == release
      (prediction.acquisitions === prediction.releases) === (balanced === true)
    );
    check(
      'branch_count_gte_1',
      prediction.branchCount,
      '>= 1', // process_data has if-not-data branch
      prediction.branchCount >= 1
    );
    const threads = observation.facts['THREADS_SPAWNED'] as number | undefined;
    if (threads !== undefined) {
      // parallelFetch creates threads — topology should see async or
      // at minimum multiple functions handling concurrency
      check(
        'concurrency_signal',
        prediction.functionCount,
        '>= 3', // fetch + process_data + parallel_fetch + worker + main
        prediction.functionCount >= 3
      );
    }
  }

  // ── Rust-specific ──
  if (lang === 'rust') {
    const errors = observation.facts['ERRORS'] as number | undefined;
    if (errors !== undefined) {
      check(
        'error_paths_gte_1',
        prediction.branchCount,
        '>= 1',
        prediction.branchCount >= 1
      );
    }
    check(
      'no_extraction_errors',
      prediction.diagnostics.length,
      0,
      prediction.diagnostics.length === 0
    );
  }

  // ── Go-specific ──
  if (lang === 'go') {
    const goroutines = observation.facts['GOROUTINES_SPAWNED'] as
      | number
      | undefined;
    if (goroutines !== undefined) {
      // parallelFetch spawns goroutines — topology should see function nodes for goroutine body
      check(
        'goroutine_fn_detected',
        prediction.functionCount,
        '>= 3', // fetch + processData + parallelFetch + closure + main
        prediction.functionCount >= 3
      );
    }
    check(
      'branch_count_gte_1',
      prediction.branchCount,
      '>= 1',
      prediction.branchCount >= 1
    );
  }

  // ── Ghost mass: no leak debt (resources are balanced in fixtures) ──
  check(
    'no_leak_debt',
    prediction.totalLeakDebt,
    0,
    prediction.totalLeakDebt === 0
  );

  return matches;
}

// ── Benchmark cases ───────────────────────────────────────────────────────

const CASES = [
  {
    name: 'Python resource + threads',
    lang: 'python',
    source: PYTHON_FIXTURE,
    ext: 'py',
    fileName: 'gnosis_native_bm.py',
  },
  {
    name: 'Rust sequential + error handling',
    lang: 'rust',
    source: RUST_FIXTURE,
    ext: 'rs',
    fileName: 'gnosis_native_bm.rs',
  },
  {
    name: 'Go goroutine fan-out + WaitGroup',
    lang: 'go',
    source: GO_FIXTURE,
    ext: 'go',
    fileName: 'gnosis_native_bm.go',
  },
] as const;

// ── Public benchmark runner ────────────────────────────────────────────────

export interface NativeRuntimeBenchmarkReport {
  results: NativeRuntimeResult[];
  overallPrecision: number;
  totalChecks: number;
  passedChecks: number;
  runtimesTestedMs: Record<string, number>;
  extractionMs: Record<string, number>;
}

/**
 * Handles the gnosis run Native Runtime Benchmark workflow.
 */
export async function runNativeRuntimeBenchmark(): Promise<NativeRuntimeBenchmarkReport> {
  const results: NativeRuntimeResult[] = [];

  for (const c of CASES) {
    // 1. Observe native runtime (sync — native exec is blocking)
    const observation = runNative(c.lang, c.source, c.ext);

    // 2. Extract topology via gnosis-polyglot binary (async)
    const { prediction, extractionMs } = await extractTopology(
      c.lang,
      c.source,
      c.fileName
    );

    // 3. Score predictions against observations
    const matches = matchPredictions(c.lang, prediction, observation);
    const passed = matches.filter((m) => m.pass).length;
    const precisionScore = matches.length > 0 ? passed / matches.length : 0;

    results.push({
      name: c.name,
      language: c.lang,
      observation,
      prediction,
      matches,
      precisionScore,
      observationMs: observation.wallClockMs,
      extractionMs,
    });
  }

  const totalChecks = results.reduce((s, r) => s + r.matches.length, 0);
  const passedChecks = results.reduce(
    (s, r) => s + r.matches.filter((m) => m.pass).length,
    0
  );
  const overallPrecision = totalChecks > 0 ? passedChecks / totalChecks : 0;

  return {
    results,
    overallPrecision,
    totalChecks,
    passedChecks,
    runtimesTestedMs: Object.fromEntries(
      results.map((r) => [r.language, r.observation.wallClockMs])
    ),
    extractionMs: Object.fromEntries(
      results.map((r) => [r.language, r.extractionMs])
    ),
  };
}

// ── Reporting ─────────────────────────────────────────────────────────────

/**
 * Handles the gnosis print Native Runtime Benchmark Report workflow.
 */
export function printNativeRuntimeBenchmarkReport(
  report: NativeRuntimeBenchmarkReport
): void {
  console.log(
    '═══════════════════════════════════════════════════════════════════'
  );
  console.log(
    ' Native Runtime Benchmark — Static Prediction vs. Empirical Reality'
  );
  console.log(
    '═══════════════════════════════════════════════════════════════════'
  );

  for (const r of report.results) {
    const icon = r.observation.exitCode === 0 ? '✓' : '✗';
    const pct = (r.precisionScore * 100).toFixed(0);
    const passed = r.matches.filter((m) => m.pass).length;
    console.log(`\n── ${icon} ${r.name} [${r.language}]`);
    console.log(
      `   Native runtime: ${r.observationMs.toFixed(
        0
      )}ms   Topology extraction: ${r.extractionMs.toFixed(0)}ms`
    );
    console.log(
      `   Functions: ${r.prediction.functionCount}   Branches: ${r.prediction.branchCount}   Returns: ${r.prediction.returnPaths}`
    );
    console.log(
      `   Acquires: ${r.prediction.acquisitions}   Releases: ${r.prediction.releases}   LeakDebt: ${r.prediction.totalLeakDebt}`
    );
    if (r.prediction.diagnostics.length > 0) {
      for (const d of r.prediction.diagnostics) console.log(`   WARNING: ${d}`);
    }
    console.log(`   Precision: ${passed}/${r.matches.length} checks (${pct}%)`);
    for (const m of r.matches) {
      const tick = m.pass ? '  ✓' : '  ✗';
      console.log(
        `${tick} ${m.property.padEnd(30)} predicted=${String(
          m.predicted
        ).padEnd(10)} observed=${m.observed}`
      );
    }
  }

  console.log(
    '\n── Overall ──────────────────────────────────────────────────────────'
  );
  console.log(
    `   Languages: ${report.results.map((r) => r.language).join(', ')}`
  );
  console.log(
    `   Checks passed: ${report.passedChecks} / ${report.totalChecks}`
  );
  console.log(`   Precision: ${(report.overallPrecision * 100).toFixed(1)}%`);

  const wallTimes = Object.entries(report.runtimesTestedMs)
    .map(([l, ms]) => `${l}: ${ms.toFixed(0)}ms`)
    .join('  ');
  const extractTimes = Object.entries(report.extractionMs)
    .map(([l, ms]) => `${l}: ${ms.toFixed(0)}ms`)
    .join('  ');
  console.log(`   Wall-clock (native):      ${wallTimes}`);
  console.log(`   Wall-clock (extraction):  ${extractTimes}`);
}

// ── CLI entry point ───────────────────────────────────────────────────────

const report = await runNativeRuntimeBenchmark();
printNativeRuntimeBenchmarkReport(report);
