// @a0n/.ignored_gnosis/src/benchmarks/near-control-sweep-benchmark.ts
import {
  buildGnosisFoldBoundaryRegimeSweepReport,
  makeDefaultRegimeSweepConfig,
  type GnosisFoldBoundaryRegimeSweepReport,
  type RegimeSweepFamilyReport,
  type RegimeSweepRunConfig,
} from './regime-sweep-benchmark';

export interface NearControlFamilyReport extends RegimeSweepFamilyReport {
  readonly lastParityRegimeValue: number | null;
  readonly nearControlDivergenceRecovered: boolean;
}

export interface GnosisNearControlSweepReport {
  readonly label: 'gnosis-near-control-sweep-v1';
  readonly config: GnosisFoldBoundaryRegimeSweepReport['config'];
  readonly affine: NearControlFamilyReport;
  readonly routed: NearControlFamilyReport;
  readonly predictedNearControlRecovered: boolean;
}

function findLastParityRegimeValue(
  family: RegimeSweepFamilyReport
): number | null {
  let lastParityRegimeValue: number | null = null;
  for (const point of family.points) {
    if (
      point.linearAdvantageEvalMeanSquaredError <= family.separationThreshold
    ) {
      lastParityRegimeValue = point.regimeValue;
    }
  }
  return lastParityRegimeValue;
}

function summarizeNearControlFamily(
  family: RegimeSweepFamilyReport
): NearControlFamilyReport {
  const lastParityRegimeValue = findLastParityRegimeValue(family);
  return {
    ...family,
    lastParityRegimeValue,
    nearControlDivergenceRecovered:
      family.firstSeparatedRegimeValue !== null &&
      lastParityRegimeValue !== null &&
      family.firstSeparatedRegimeValue > lastParityRegimeValue,
  };
}

/**
 * Handles the gnosis make Near Control Sweep Config workflow.
 */
export function makeNearControlSweepConfig(): RegimeSweepRunConfig {
  const base = makeDefaultRegimeSweepConfig();
  return {
    affine: {
      ...base.affine,
      regimeValues: [0, 0.05, 0.1, 0.15, 0.2, 0.25, 0.3, 0.35, 0.4, 0.5],
    },
    routed: {
      ...base.routed,
      regimeValues: [0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.55, 0.6, 0.65, 0.75],
    },
  };
}

/**
 * Handles the gnosis run Gnosis Near Control Sweep workflow.
 */
export async function runGnosisNearControlSweep(): Promise<GnosisNearControlSweepReport> {
  const baseReport = await buildGnosisFoldBoundaryRegimeSweepReport(
    makeNearControlSweepConfig(),
    'gnosis-near-control-sweep-v1'
  );
  const affine = summarizeNearControlFamily(baseReport.affine);
  const routed = summarizeNearControlFamily(baseReport.routed);

  return {
    label: 'gnosis-near-control-sweep-v1',
    config: baseReport.config,
    affine,
    routed,
    predictedNearControlRecovered:
      affine.nearControlDivergenceRecovered &&
      routed.nearControlDivergenceRecovered &&
      affine.phaseBoundaryRecovered &&
      routed.phaseBoundaryRecovered,
  };
}

function renderFamilyMarkdown(family: NearControlFamilyReport): string {
  const lines: string[] = [];
  lines.push(`## ${family.title}`);
  lines.push('');
  lines.push(`- Description: ${family.description}`);
  lines.push(`- Sweep dimension: ${family.sweepDimension}`);
  lines.push(
    `- Last parity regime value: \`${
      family.lastParityRegimeValue === null
        ? 'none'
        : family.lastParityRegimeValue.toFixed(2)
    }\``
  );
  lines.push(
    `- First separated regime value: \`${
      family.firstSeparatedRegimeValue === null
        ? 'none'
        : family.firstSeparatedRegimeValue.toFixed(2)
    }\``
  );
  lines.push(
    `- Near-control divergence recovered: \`${
      family.nearControlDivergenceRecovered ? 'yes' : 'no'
    }\``
  );
  lines.push('');
  lines.push(
    '| Regime | Linear advantage (eval MSE) | Linear exact advantage | Best nonlinear |'
  );
  lines.push('| ---: | ---: | ---: | --- |');
  for (const point of family.points) {
    lines.push(
      `| ${point.regimeValue.toFixed(
        2
      )} | ${point.linearAdvantageEvalMeanSquaredError.toFixed(
        3
      )} | ${point.linearAdvantageExactWithinToleranceFraction.toFixed(
        3
      )} | \`${point.bestNonlinearStrategy}\` |`
    );
  }
  lines.push('');
  return lines.join('\n');
}

/**
 * Handles the gnosis render Gnosis Near Control Sweep Markdown workflow.
 */
export function renderGnosisNearControlSweepMarkdown(
  report: GnosisNearControlSweepReport
): string {
  const lines: string[] = [];
  lines.push('# Gnosis Near-Control Sweep');
  lines.push('');
  lines.push(`- Label: \`${report.label}\``);
  lines.push(
    `- Predicted near-control recovery: \`${
      report.predictedNearControlRecovered ? 'yes' : 'no'
    }\``
  );
  lines.push(
    `- Affine regime values: ${report.affine.regimeValues
      .map((value) => `\`${value.toFixed(2)}\``)
      .join(', ')}`
  );
  lines.push(
    `- Routed regime values: ${report.routed.regimeValues
      .map((value) => `\`${value.toFixed(2)}\``)
      .join(', ')}`
  );
  lines.push('');
  lines.push(renderFamilyMarkdown(report.affine));
  lines.push('');
  lines.push(renderFamilyMarkdown(report.routed));
  lines.push('');
  lines.push(
    'Interpretation: this near-control sweep zooms the control end of the boundary, showing how long parity persists before linear recombination opens a measurable advantage as additive demand rises.'
  );
  lines.push('');
  return `${lines.join('\n')}\n`;
}
