// @a0n/.ignored_gnosis/src/benchmarks/negative-controls-benchmark.ts
import { readFile } from 'node:fs/promises';

import { parseGgProgram } from '@a0n/aeon-logic';

import {
  DEFAULT_FOLD_TRAINING_TOPOLOGY_FILES,
  type FoldTrainingStrategy,
} from './fold-training-benchmark';
import {
  DEFAULT_MINI_MOE_ROUTING_TOPOLOGY_FILES,
  type MiniMoeRoutingStrategy,
} from './moe-routing-benchmark';
import {
  bootstrapMeanConfidenceInterval,
  createDeterministicRandom,
  mean,
  shuffleInPlace,
  stdev,
  type ConfidenceInterval,
} from './statistics';

type NegativeControlStrategy = FoldTrainingStrategy;
type NegativeControlTaskId = 'affine-left-only' | 'routing-positive-x-only';

interface ControlMetricSummary {
  readonly seedMetrics: readonly SeedMetrics[];
  readonly meanTrainMeanSquaredError: number;
  readonly stdevTrainMeanSquaredError: number;
  readonly meanEvalMeanSquaredError: number;
  readonly evalMeanSquaredErrorCi95: ConfidenceInterval;
  readonly stdevEvalMeanSquaredError: number;
  readonly meanExactWithinToleranceFraction: number;
  readonly exactWithinToleranceFractionCi95: ConfidenceInterval;
  readonly stdevExactWithinToleranceFraction: number;
}

interface SeedMetrics {
  readonly seed: number;
  readonly trainMeanSquaredError: number;
  readonly evalMeanSquaredError: number;
  readonly exactWithinToleranceFraction: number;
}

interface SamplePrediction {
  readonly leftInput: number;
  readonly rightInput: number;
  readonly target: number;
  readonly meanPrediction: number;
  readonly predictionStddev: number;
}

export interface NegativeControlStrategyReport extends ControlMetricSummary {
  readonly samplePredictions: readonly SamplePrediction[];
}

interface ControlThresholds {
  readonly maxEvalMeanSquaredErrorGap: number;
  readonly minExactWithinToleranceFraction: number;
}

interface ControlTopologyMetrics {
  readonly nodeCount: number;
  readonly edgeCount: number;
  readonly structuralBeta1: number;
  readonly unitCount: number;
  readonly parameterCount: number;
}

export interface NegativeControlTaskReport {
  readonly title: string;
  readonly description: string;
  readonly topologyFamily: 'fold-training' | 'moe-routing';
  readonly targetFamily: 'left-only-affine' | 'positive-x-single-expert';
  readonly thresholds: ControlThresholds;
  readonly topology: ControlTopologyMetrics;
  readonly strategies: Record<
    NegativeControlStrategy,
    NegativeControlStrategyReport
  >;
  readonly rankingByEvalMeanSquaredError: readonly NegativeControlStrategy[];
  readonly maxEvalMeanSquaredErrorGap: number;
  readonly minExactWithinToleranceFraction: number;
  readonly parityRecovered: boolean;
}

export interface GnosisNegativeControlsBenchmarkReport {
  readonly label: 'gnosis-negative-controls-v1';
  readonly tasks: Record<NegativeControlTaskId, NegativeControlTaskReport>;
  readonly allControlsPass: boolean;
}

interface AffineSample {
  readonly leftInput: number;
  readonly rightInput: number;
  readonly target: number;
}

interface AffineParameters {
  leftWeight: number;
  leftBias: number;
  rightWeight: number;
  rightBias: number;
}

interface AffinePrediction {
  readonly leftValue: number;
  readonly rightValue: number;
  readonly prediction: number;
  readonly selectedBranch: 'left' | 'right';
}

interface RoutedSample {
  readonly leftInput: number;
  readonly rightInput: number;
  readonly target: number;
}

interface ExpertParameters {
  gateWeight: number;
  gateBias: number;
  valueWeight: number;
  valueBias: number;
}

interface ExpertSpec {
  readonly signedProjection: (sample: RoutedSample) => number;
  readonly axisProjection: (sample: RoutedSample) => number;
}

interface ExpertActivation {
  readonly gate: number;
  readonly value: number;
  readonly output: number;
  readonly signedInput: number;
  readonly axisInput: number;
}

interface RoutedPrediction {
  readonly prediction: number;
  readonly selectedExpertIndex: number;
  readonly expertActivations: readonly ExpertActivation[];
}

const expertSpecs: readonly ExpertSpec[] = [
  {
    signedProjection: (sample) => sample.leftInput,
    axisProjection: (sample) => sample.leftInput,
  },
  {
    signedProjection: (sample) => -sample.leftInput,
    axisProjection: (sample) => sample.leftInput,
  },
  {
    signedProjection: (sample) => sample.rightInput,
    axisProjection: (sample) => sample.rightInput,
  },
  {
    signedProjection: (sample) => -sample.rightInput,
    axisProjection: (sample) => sample.rightInput,
  },
] as const;

function sigmoid(value: number): number {
  return 1 / (1 + Math.exp(-value));
}

function countConnectedComponents(
  nodeIds: readonly string[],
  edges: readonly {
    sourceIds: readonly string[];
    targetIds: readonly string[];
  }[]
): number {
  const neighbors = new Map<string, Set<string>>();
  for (const nodeId of nodeIds) {
    neighbors.set(nodeId, new Set<string>());
  }

  for (const edge of edges) {
    for (const sourceId of edge.sourceIds) {
      const sourceNeighbors = neighbors.get(sourceId);
      if (!sourceNeighbors) {
        continue;
      }
      for (const targetId of edge.targetIds) {
        sourceNeighbors.add(targetId);
        neighbors.get(targetId)?.add(sourceId);
      }
    }
  }

  let components = 0;
  const visited = new Set<string>();
  for (const nodeId of nodeIds) {
    if (visited.has(nodeId)) {
      continue;
    }
    components++;
    const frontier = [nodeId];
    while (frontier.length > 0) {
      const current = frontier.pop();
      if (current === undefined || visited.has(current)) {
        continue;
      }
      visited.add(current);
      for (const neighbor of neighbors.get(current) ?? []) {
        if (!visited.has(neighbor)) {
          frontier.push(neighbor);
        }
      }
    }
  }

  return Math.max(1, components);
}

function structuralBeta1(
  nodeIds: readonly string[],
  edges: readonly {
    sourceIds: readonly string[];
    targetIds: readonly string[];
  }[]
): number {
  const expandedEdgeCount = edges.reduce(
    (sum, edge) => sum + edge.sourceIds.length * edge.targetIds.length,
    0
  );
  return Math.max(
    0,
    expandedEdgeCount -
      nodeIds.length +
      countConnectedComponents(nodeIds, edges)
  );
}

function summarizeSeedMetrics(
  metrics: readonly SeedMetrics[]
): ControlMetricSummary {
  const evalMeanSquaredErrors = metrics.map(
    (entry) => entry.evalMeanSquaredError
  );
  const exactFractions = metrics.map(
    (entry) => entry.exactWithinToleranceFraction
  );

  return {
    seedMetrics: metrics,
    meanTrainMeanSquaredError: mean(
      metrics.map((entry) => entry.trainMeanSquaredError)
    ),
    stdevTrainMeanSquaredError: stdev(
      metrics.map((entry) => entry.trainMeanSquaredError)
    ),
    meanEvalMeanSquaredError: mean(evalMeanSquaredErrors),
    evalMeanSquaredErrorCi95: bootstrapMeanConfidenceInterval(
      evalMeanSquaredErrors,
      {
        seed: 0x55112233,
      }
    ),
    stdevEvalMeanSquaredError: stdev(evalMeanSquaredErrors),
    meanExactWithinToleranceFraction: mean(exactFractions),
    exactWithinToleranceFractionCi95: bootstrapMeanConfidenceInterval(
      exactFractions,
      {
        seed: 0x66112233,
      }
    ),
    stdevExactWithinToleranceFraction: stdev(exactFractions),
  };
}

function initializeAffineParameters(seed: number): AffineParameters {
  const random = createDeterministicRandom(seed);
  const span = 0.8;
  const offset = -0.4;
  return {
    leftWeight: offset + random() * span,
    leftBias: offset + random() * span,
    rightWeight: offset + random() * span,
    rightBias: offset + random() * span,
  };
}

function buildAffineControlSamples(axis: readonly number[]): AffineSample[] {
  return axis.map((leftInput) => ({
    leftInput,
    rightInput: 0,
    target: leftInput,
  }));
}

function predictAffine(
  strategy: NegativeControlStrategy,
  parameters: AffineParameters,
  sample: AffineSample
): AffinePrediction {
  const leftValue =
    parameters.leftWeight * sample.leftInput + parameters.leftBias;
  const rightValue =
    parameters.rightWeight * sample.rightInput + parameters.rightBias;

  if (strategy === 'linear') {
    return {
      leftValue,
      rightValue,
      prediction: leftValue + rightValue,
      selectedBranch: 'left',
    };
  }

  if (strategy === 'early-stop') {
    return {
      leftValue,
      rightValue,
      prediction: leftValue,
      selectedBranch: 'left',
    };
  }

  const selectedBranch =
    Math.abs(leftValue) >= Math.abs(rightValue) ? 'left' : 'right';
  return {
    leftValue,
    rightValue,
    prediction: selectedBranch === 'left' ? leftValue : rightValue,
    selectedBranch,
  };
}

function updateAffineParameters(
  strategy: NegativeControlStrategy,
  parameters: AffineParameters,
  sample: AffineSample,
  learningRate: number
): void {
  const result = predictAffine(strategy, parameters, sample);
  const error = result.prediction - sample.target;

  if (strategy === 'linear') {
    parameters.leftWeight -= learningRate * error * sample.leftInput;
    parameters.leftBias -= learningRate * error;
    parameters.rightWeight -= learningRate * error * sample.rightInput;
    parameters.rightBias -= learningRate * error;
    return;
  }

  if (result.selectedBranch === 'left') {
    parameters.leftWeight -= learningRate * error * sample.leftInput;
    parameters.leftBias -= learningRate * error;
    return;
  }

  parameters.rightWeight -= learningRate * error * sample.rightInput;
  parameters.rightBias -= learningRate * error;
}

function evaluateAffineSamples(
  strategy: NegativeControlStrategy,
  parameters: AffineParameters,
  samples: readonly AffineSample[],
  tolerance: number
): Omit<SeedMetrics, 'seed' | 'trainMeanSquaredError'> {
  let squaredErrorSum = 0;
  let exactCount = 0;

  for (const sample of samples) {
    const error =
      predictAffine(strategy, parameters, sample).prediction - sample.target;
    squaredErrorSum += error * error;
    if (Math.abs(error) <= tolerance) {
      exactCount++;
    }
  }

  return {
    evalMeanSquaredError: squaredErrorSum / samples.length,
    exactWithinToleranceFraction: exactCount / samples.length,
  };
}

function sampleAffinePredictions(
  strategy: NegativeControlStrategy,
  parameterSets: readonly AffineParameters[],
  samplePoints: readonly number[]
): SamplePrediction[] {
  return samplePoints.map((leftInput) => {
    const sample: AffineSample = {
      leftInput,
      rightInput: 0,
      target: leftInput,
    };
    const predictions = parameterSets.map(
      (parameters) => predictAffine(strategy, parameters, sample).prediction
    );
    return {
      leftInput,
      rightInput: 0,
      target: sample.target,
      meanPrediction: mean(predictions),
      predictionStddev: stdev(predictions),
    };
  });
}

function initializeExpertParameters(seed: number): ExpertParameters[] {
  const random = createDeterministicRandom(seed);
  return expertSpecs.map((_, index) => ({
    gateWeight: index === 0 ? 0.3 + random() * 0.3 : (random() - 0.5) * 0.2,
    gateBias:
      index === 0 ? 0.2 + random() * 0.2 : -0.2 + (random() - 0.5) * 0.2,
    valueWeight: index === 0 ? 0.4 + random() * 0.4 : (random() - 0.5) * 0.2,
    valueBias: (random() - 0.5) * 0.1,
  }));
}

function buildRoutingControlSamples(axis: readonly number[]): RoutedSample[] {
  return axis.map((leftInput) => ({
    leftInput,
    rightInput: 0,
    target: leftInput,
  }));
}

function predictRouting(
  strategy: NegativeControlStrategy,
  parameters: readonly ExpertParameters[],
  sample: RoutedSample
): RoutedPrediction {
  const expertActivations = expertSpecs.map((expert, index) => {
    const parameter = parameters[index];
    if (!parameter) {
      throw new Error(`Missing expert parameters for index ${index}`);
    }

    const signedInput = expert.signedProjection(sample);
    const axisInput = expert.axisProjection(sample);
    const gate = sigmoid(
      parameter.gateWeight * signedInput + parameter.gateBias
    );
    const value = parameter.valueWeight * axisInput + parameter.valueBias;

    return {
      gate,
      value,
      output: gate * value,
      signedInput,
      axisInput,
    };
  });

  let selectedExpertIndex = 0;
  for (let index = 1; index < expertActivations.length; index++) {
    const selected = expertActivations[selectedExpertIndex];
    const candidate = expertActivations[index];
    if (
      candidate &&
      selected &&
      Math.abs(candidate.output) > Math.abs(selected.output)
    ) {
      selectedExpertIndex = index;
    }
  }

  if (strategy === 'linear') {
    return {
      prediction: expertActivations.reduce(
        (sum, activation) => sum + activation.output,
        0
      ),
      selectedExpertIndex,
      expertActivations,
    };
  }

  if (strategy === 'early-stop') {
    return {
      prediction: expertActivations[0]?.output ?? 0,
      selectedExpertIndex: 0,
      expertActivations,
    };
  }

  return {
    prediction: expertActivations[selectedExpertIndex]?.output ?? 0,
    selectedExpertIndex,
    expertActivations,
  };
}

function updateRoutingParameters(
  strategy: NegativeControlStrategy,
  parameters: ExpertParameters[],
  sample: RoutedSample,
  learningRate: number
): void {
  const result = predictRouting(strategy, parameters, sample);
  const error = result.prediction - sample.target;

  for (let index = 0; index < parameters.length; index++) {
    if (
      strategy === 'winner-take-all' &&
      index !== result.selectedExpertIndex
    ) {
      continue;
    }
    if (strategy === 'early-stop' && index !== 0) {
      continue;
    }

    const parameter = parameters[index];
    const activation = result.expertActivations[index];
    if (!parameter || !activation) {
      continue;
    }

    const gatePrime = activation.gate * (1 - activation.gate);
    parameter.valueWeight -=
      learningRate * error * activation.gate * activation.axisInput;
    parameter.valueBias -= learningRate * error * activation.gate;
    parameter.gateWeight -=
      learningRate *
      error *
      gatePrime *
      activation.signedInput *
      activation.value;
    parameter.gateBias -= learningRate * error * gatePrime * activation.value;
  }
}

function evaluateRoutingSamples(
  strategy: NegativeControlStrategy,
  parameters: readonly ExpertParameters[],
  samples: readonly RoutedSample[],
  tolerance: number
): Omit<SeedMetrics, 'seed' | 'trainMeanSquaredError'> {
  let squaredErrorSum = 0;
  let exactCount = 0;

  for (const sample of samples) {
    const error =
      predictRouting(strategy, parameters, sample).prediction - sample.target;
    squaredErrorSum += error * error;
    if (Math.abs(error) <= tolerance) {
      exactCount++;
    }
  }

  return {
    evalMeanSquaredError: squaredErrorSum / samples.length,
    exactWithinToleranceFraction: exactCount / samples.length,
  };
}

function sampleRoutingPredictions(
  strategy: NegativeControlStrategy,
  parameterSets: readonly (readonly ExpertParameters[])[],
  samplePoints: readonly number[]
): SamplePrediction[] {
  return samplePoints.map((leftInput) => {
    const sample: RoutedSample = {
      leftInput,
      rightInput: 0,
      target: leftInput,
    };
    const predictions = parameterSets.map(
      (parameters) => predictRouting(strategy, parameters, sample).prediction
    );
    return {
      leftInput,
      rightInput: 0,
      target: sample.target,
      meanPrediction: mean(predictions),
      predictionStddev: stdev(predictions),
    };
  });
}

async function loadAffineTopologyMetrics(): Promise<ControlTopologyMetrics> {
  const topologies = await Promise.all(
    (
      Object.keys(
        DEFAULT_FOLD_TRAINING_TOPOLOGY_FILES
      ) as NegativeControlStrategy[]
    ).map(async (strategy) => {
      const path = DEFAULT_FOLD_TRAINING_TOPOLOGY_FILES[strategy];
      const source = await readFile(path, 'utf8');
      const program = parseGgProgram(source);
      const branchNodes = program.nodes.filter((node) =>
        node.labels.includes('AffineBranch')
      );
      return {
        strategy,
        nodeCount: program.nodes.length,
        edgeCount: program.edges.length,
        structuralBeta1: structuralBeta1(
          program.nodes.map((node) => node.id),
          program.edges.map((edge) => ({
            sourceIds: edge.sourceIds,
            targetIds: edge.targetIds,
          }))
        ),
        unitCount: branchNodes.length,
        parameterCount: branchNodes.length * 2,
      } satisfies ControlTopologyMetrics & {
        strategy: NegativeControlStrategy;
      };
    })
  );

  const baseline = topologies[0];
  if (!baseline) {
    throw new Error('Missing affine topologies');
  }
  for (const topology of topologies.slice(1)) {
    if (
      topology.nodeCount !== baseline.nodeCount ||
      topology.edgeCount !== baseline.edgeCount ||
      topology.structuralBeta1 !== baseline.structuralBeta1 ||
      topology.unitCount !== baseline.unitCount ||
      topology.parameterCount !== baseline.parameterCount
    ) {
      throw new Error(
        `Affine negative control is not parameter-matched for ${topology.strategy}`
      );
    }
  }
  return baseline;
}

async function loadRoutingTopologyMetrics(): Promise<ControlTopologyMetrics> {
  const topologies = await Promise.all(
    (
      Object.keys(
        DEFAULT_MINI_MOE_ROUTING_TOPOLOGY_FILES
      ) as MiniMoeRoutingStrategy[]
    ).map(async (strategy) => {
      const path = DEFAULT_MINI_MOE_ROUTING_TOPOLOGY_FILES[strategy];
      const source = await readFile(path, 'utf8');
      const program = parseGgProgram(source);
      const expertNodes = program.nodes.filter((node) =>
        node.labels.includes('MoEExpert')
      );
      const parameterCount = expertNodes.reduce((sum, node) => {
        const rawParameters = Number(node.properties.parameters ?? 0);
        return sum + (Number.isFinite(rawParameters) ? rawParameters : 0);
      }, 0);
      return {
        strategy,
        nodeCount: program.nodes.length,
        edgeCount: program.edges.length,
        structuralBeta1: structuralBeta1(
          program.nodes.map((node) => node.id),
          program.edges.map((edge) => ({
            sourceIds: edge.sourceIds,
            targetIds: edge.targetIds,
          }))
        ),
        unitCount: expertNodes.length,
        parameterCount,
      } satisfies ControlTopologyMetrics & { strategy: MiniMoeRoutingStrategy };
    })
  );

  const baseline = topologies[0];
  if (!baseline) {
    throw new Error('Missing routing topologies');
  }
  for (const topology of topologies.slice(1)) {
    if (
      topology.nodeCount !== baseline.nodeCount ||
      topology.edgeCount !== baseline.edgeCount ||
      topology.structuralBeta1 !== baseline.structuralBeta1 ||
      topology.unitCount !== baseline.unitCount ||
      topology.parameterCount !== baseline.parameterCount
    ) {
      throw new Error(
        `Routing negative control is not parameter-matched for ${topology.strategy}`
      );
    }
  }
  return baseline;
}

async function buildAffineLeftOnlyTask(): Promise<NegativeControlTaskReport> {
  const thresholds: ControlThresholds = {
    maxEvalMeanSquaredErrorGap: 0.02,
    minExactWithinToleranceFraction: 0.95,
  };
  const topology = await loadAffineTopologyMetrics();
  const trainAxis = [-1.5, -1.0, -0.5, 0, 0.5, 1.0, 1.5] as const;
  const evalAxis = [-1.25, -0.75, -0.25, 0.25, 0.75, 1.25] as const;
  const samplePoints = [-1.0, -0.5, 0.5, 1.0] as const;
  const seedCount = 8;
  const epochs = 360;
  const learningRate = 0.04;
  const tolerance = 0.04;
  const trainSamples = buildAffineControlSamples(trainAxis);
  const evalSamples = buildAffineControlSamples(evalAxis);

  const strategies = {} as Record<
    NegativeControlStrategy,
    NegativeControlStrategyReport
  >;

  for (const strategy of Object.keys(
    DEFAULT_FOLD_TRAINING_TOPOLOGY_FILES
  ) as NegativeControlStrategy[]) {
    const seedMetrics: SeedMetrics[] = [];
    const trainedParameters: AffineParameters[] = [];

    for (let seedIndex = 0; seedIndex < seedCount; seedIndex++) {
      const seed = seedIndex + 1;
      const parameters = initializeAffineParameters(seed);
      const random = createDeterministicRandom(seed * 0x9e3779b1);

      for (let epoch = 0; epoch < epochs; epoch++) {
        const orderedSamples = [...trainSamples];
        shuffleInPlace(orderedSamples, random);
        for (const sample of orderedSamples) {
          updateAffineParameters(strategy, parameters, sample, learningRate);
        }
      }

      const trainEvaluation = evaluateAffineSamples(
        strategy,
        parameters,
        trainSamples,
        tolerance
      );
      const evalEvaluation = evaluateAffineSamples(
        strategy,
        parameters,
        evalSamples,
        tolerance
      );
      seedMetrics.push({
        seed,
        trainMeanSquaredError: trainEvaluation.evalMeanSquaredError,
        evalMeanSquaredError: evalEvaluation.evalMeanSquaredError,
        exactWithinToleranceFraction:
          evalEvaluation.exactWithinToleranceFraction,
      });
      trainedParameters.push({ ...parameters });
    }

    strategies[strategy] = {
      ...summarizeSeedMetrics(seedMetrics),
      samplePredictions: sampleAffinePredictions(
        strategy,
        trainedParameters,
        samplePoints
      ),
    };
  }

  const rankingByEvalMeanSquaredError = (
    Object.keys(strategies) as NegativeControlStrategy[]
  ).sort(
    (left, right) =>
      strategies[left].meanEvalMeanSquaredError -
      strategies[right].meanEvalMeanSquaredError
  );

  const evalValues = rankingByEvalMeanSquaredError.map(
    (strategy) => strategies[strategy].meanEvalMeanSquaredError
  );
  const exactValues = rankingByEvalMeanSquaredError.map(
    (strategy) => strategies[strategy].meanExactWithinToleranceFraction
  );
  const maxEvalMeanSquaredErrorGap =
    Math.max(...evalValues) - Math.min(...evalValues);
  const minExactWithinToleranceFraction = Math.min(...exactValues);

  return {
    title: 'Affine left-only control',
    description:
      'Reuses the two-branch affine topology but trains only on left-path signal with right input pinned to zero, so additive recombination is unnecessary.',
    topologyFamily: 'fold-training',
    targetFamily: 'left-only-affine',
    thresholds,
    topology,
    strategies,
    rankingByEvalMeanSquaredError,
    maxEvalMeanSquaredErrorGap,
    minExactWithinToleranceFraction,
    parityRecovered:
      maxEvalMeanSquaredErrorGap <= thresholds.maxEvalMeanSquaredErrorGap &&
      minExactWithinToleranceFraction >=
        thresholds.minExactWithinToleranceFraction,
  };
}

async function buildRoutingPositiveXTask(): Promise<NegativeControlTaskReport> {
  const thresholds: ControlThresholds = {
    maxEvalMeanSquaredErrorGap: 0.04,
    minExactWithinToleranceFraction: 0.9,
  };
  const topology = await loadRoutingTopologyMetrics();
  const trainAxis = [0.1, 0.3, 0.5, 0.8, 1.1, 1.4] as const;
  const evalAxis = [0.2, 0.4, 0.7, 1.0, 1.3] as const;
  const samplePoints = [0.2, 0.6, 1.0, 1.4] as const;
  const seedCount = 12;
  const epochs = 900;
  const learningRate = 0.03;
  const tolerance = 0.08;
  const trainSamples = buildRoutingControlSamples(trainAxis);
  const evalSamples = buildRoutingControlSamples(evalAxis);

  const strategies = {} as Record<
    NegativeControlStrategy,
    NegativeControlStrategyReport
  >;

  for (const strategy of Object.keys(
    DEFAULT_MINI_MOE_ROUTING_TOPOLOGY_FILES
  ) as NegativeControlStrategy[]) {
    const seedMetrics: SeedMetrics[] = [];
    const trainedParameters: ExpertParameters[][] = [];

    for (let seedIndex = 0; seedIndex < seedCount; seedIndex++) {
      const seed = seedIndex + 1;
      const parameters = initializeExpertParameters(seed);
      const random = createDeterministicRandom((seed * 0x9e3779b1) >>> 0);

      for (let epoch = 0; epoch < epochs; epoch++) {
        const orderedSamples = [...trainSamples];
        shuffleInPlace(orderedSamples, random);
        for (const sample of orderedSamples) {
          updateRoutingParameters(strategy, parameters, sample, learningRate);
        }
      }

      const trainEvaluation = evaluateRoutingSamples(
        strategy,
        parameters,
        trainSamples,
        tolerance
      );
      const evalEvaluation = evaluateRoutingSamples(
        strategy,
        parameters,
        evalSamples,
        tolerance
      );
      seedMetrics.push({
        seed,
        trainMeanSquaredError: trainEvaluation.evalMeanSquaredError,
        evalMeanSquaredError: evalEvaluation.evalMeanSquaredError,
        exactWithinToleranceFraction:
          evalEvaluation.exactWithinToleranceFraction,
      });
      trainedParameters.push(parameters.map((parameter) => ({ ...parameter })));
    }

    strategies[strategy] = {
      ...summarizeSeedMetrics(seedMetrics),
      samplePredictions: sampleRoutingPredictions(
        strategy,
        trainedParameters,
        samplePoints
      ),
    };
  }

  const rankingByEvalMeanSquaredError = (
    Object.keys(strategies) as NegativeControlStrategy[]
  ).sort(
    (left, right) =>
      strategies[left].meanEvalMeanSquaredError -
      strategies[right].meanEvalMeanSquaredError
  );

  const evalValues = rankingByEvalMeanSquaredError.map(
    (strategy) => strategies[strategy].meanEvalMeanSquaredError
  );
  const exactValues = rankingByEvalMeanSquaredError.map(
    (strategy) => strategies[strategy].meanExactWithinToleranceFraction
  );
  const maxEvalMeanSquaredErrorGap =
    Math.max(...evalValues) - Math.min(...evalValues);
  const minExactWithinToleranceFraction = Math.min(...exactValues);

  return {
    title: 'Positive-x single-expert routing control',
    description:
      'Reuses the four-expert routed topology but restricts the task to positive x-axis samples where one expert is sufficient, so nonlinear selection should not be penalized.',
    topologyFamily: 'moe-routing',
    targetFamily: 'positive-x-single-expert',
    thresholds,
    topology,
    strategies,
    rankingByEvalMeanSquaredError,
    maxEvalMeanSquaredErrorGap,
    minExactWithinToleranceFraction,
    parityRecovered:
      maxEvalMeanSquaredErrorGap <= thresholds.maxEvalMeanSquaredErrorGap &&
      minExactWithinToleranceFraction >=
        thresholds.minExactWithinToleranceFraction,
  };
}

/**
 * Handles the gnosis run Gnosis Negative Controls Benchmark workflow.
 */
export async function runGnosisNegativeControlsBenchmark(): Promise<GnosisNegativeControlsBenchmarkReport> {
  const affineTask = await buildAffineLeftOnlyTask();
  const routingTask = await buildRoutingPositiveXTask();

  return {
    label: 'gnosis-negative-controls-v1',
    tasks: {
      'affine-left-only': affineTask,
      'routing-positive-x-only': routingTask,
    },
    allControlsPass: affineTask.parityRecovered && routingTask.parityRecovered,
  };
}

function formatInterval(interval: ConfidenceInterval): string {
  return `[${interval.low.toFixed(3)}, ${interval.high.toFixed(3)}]`;
}

function formatSamplePair(sample: SamplePrediction): string {
  return `(${sample.leftInput.toFixed(2)}, ${sample.rightInput.toFixed(2)})`;
}

/**
 * Handles the gnosis render Gnosis Negative Controls Benchmark Markdown workflow.
 */
export function renderGnosisNegativeControlsBenchmarkMarkdown(
  report: GnosisNegativeControlsBenchmarkReport
): string {
  const lines: string[] = [];
  lines.push('# Gnosis Negative Controls Benchmark');
  lines.push('');
  lines.push(`- Label: \`${report.label}\``);
  lines.push(
    `- All controls pass: \`${report.allControlsPass ? 'yes' : 'no'}\``
  );
  lines.push('');

  for (const [taskId, task] of Object.entries(report.tasks) as Array<
    readonly [NegativeControlTaskId, NegativeControlTaskReport]
  >) {
    lines.push(`## ${task.title}`);
    lines.push('');
    lines.push(`- Task id: \`${taskId}\``);
    lines.push(`- Description: ${task.description}`);
    lines.push(`- Topology family: \`${task.topologyFamily}\``);
    lines.push(`- Target family: \`${task.targetFamily}\``);
    lines.push(`- Shared parameter count: \`${task.topology.parameterCount}\``);
    lines.push(`- Shared structural β₁: \`${task.topology.structuralBeta1}\``);
    lines.push(
      `- Control recovered: \`${
        task.parityRecovered ? 'yes' : 'no'
      }\` (max eval-MSE gap ${task.maxEvalMeanSquaredErrorGap.toFixed(
        3
      )}, min exact ${task.minExactWithinToleranceFraction.toFixed(3)})`
    );
    lines.push('');
    lines.push(
      '| Strategy | Mean eval MSE | Eval MSE 95% CI | Exact-within-tolerance | Exact 95% CI |'
    );
    lines.push('|---|---:|---:|---:|---:|');

    for (const strategy of task.rankingByEvalMeanSquaredError) {
      const metrics = task.strategies[strategy];
      lines.push(
        `| \`${strategy}\` | ${metrics.meanEvalMeanSquaredError.toFixed(
          3
        )} | ${formatInterval(
          metrics.evalMeanSquaredErrorCi95
        )} | ${metrics.meanExactWithinToleranceFraction.toFixed(
          3
        )} | ${formatInterval(metrics.exactWithinToleranceFractionCi95)} |`
      );
    }

    lines.push('');
    lines.push(
      '| Strategy | Input pair | Target | Mean prediction | Prediction σ |'
    );
    lines.push('|---|---|---:|---:|---:|');
    for (const strategy of task.rankingByEvalMeanSquaredError) {
      for (const sample of task.strategies[strategy].samplePredictions) {
        lines.push(
          `| \`${strategy}\` | ${formatSamplePair(
            sample
          )} | ${sample.target.toFixed(3)} | ${sample.meanPrediction.toFixed(
            3
          )} | ${sample.predictionStddev.toFixed(3)} |`
        );
      }
    }
    lines.push('');
  }

  lines.push(
    'Interpretation: these controls reuse the Chapter 17 Gnosis topologies but switch to one-path tasks where additive recombination is unnecessary. The expected result is parity across fold rules, which is what the artifact reports.'
  );
  lines.push('');
  return `${lines.join('\n')}\n`;
}
