// @a0n/.ignored_gnosis/src/benchmarks/pattern-horizon-benchmark.ts
import {
  compilePatternHorizonWindows,
  derivePatternHorizonCompiler,
  type PatternHorizonCompilerInput,
  type PatternHorizonPolicyId,
  type PatternHorizonWindow,
} from '../pattern-horizon-compiler';

export interface PatternHorizonBenchmarkScenario {
  readonly scenarioId: string;
  readonly label: string;
  readonly input: PatternHorizonCompilerInput;
}

export interface PatternHorizonBenchmarkScenarioResult {
  readonly scenarioId: string;
  readonly label: string;
  readonly policy: PatternHorizonPolicyId;
  readonly committed: boolean;
  readonly authorityReady: boolean;
  readonly selectedWindowId: string | null;
  readonly baselineWindowId: string | null;
  readonly baselineSelectedMachineOnly: boolean;
  readonly unsafeBaselineCommitBlocked: boolean;
  readonly machineOnlyBaselineSelectionBlocked: boolean;
  readonly sameSelectionAsBaseline: boolean;
  readonly candidateCount: number;
  readonly suppressedMachineOnlyCount: number;
  readonly suppressedOutsideHorizonCount: number;
}

export interface PatternHorizonBenchmarkReport {
  readonly label: 'gnosis-pattern-horizon-benchmark-v1';
  readonly scenarioCount: number;
  readonly compiler: {
    readonly commitRate: number;
    readonly safeStateRate: number;
    readonly hedgeShare: number;
    readonly escalationShare: number;
    readonly meanCandidateCount: number;
    readonly meanSuppressedMachineOnlyCount: number;
    readonly meanSuppressedOutsideHorizonCount: number;
  };
  readonly baseline: {
    readonly commitRate: number;
    readonly machineOnlyCommitRate: number;
  };
  readonly comparisons: {
    readonly unsafeBaselineCommitsBlocked: number;
    readonly machineOnlyBaselineSelectionsBlocked: number;
    readonly sameSelectionRate: number;
    readonly divergentSelectionRate: number;
  };
  readonly scenarios: readonly PatternHorizonBenchmarkScenarioResult[];
}

function mean(values: readonly number[]): number {
  if (values.length === 0) {
    return 0;
  }
  return values.reduce((sum, value) => sum + value, 0) / values.length;
}

function selectNaiveBaselineWindow(
  input: PatternHorizonCompilerInput,
  phaseBeat: number,
  horizonBeats: number,
  catchToleranceBeats: number
): PatternHorizonWindow | null {
  const baselineBundle = derivePatternHorizonCompiler({
    ...input,
    allowMachineOnly: true,
  });
  const baselineWindows = compilePatternHorizonWindows(
    baselineBundle.orbit,
    phaseBeat,
    horizonBeats,
    catchToleranceBeats,
    true
  );

  return (
    baselineWindows.find((window) => window.admissible) ??
    baselineBundle.windows.find((window) => window.admissible) ??
    null
  );
}

function buildScenarioResult(
  scenario: PatternHorizonBenchmarkScenario
): PatternHorizonBenchmarkScenarioResult {
  const compiler = derivePatternHorizonCompiler(scenario.input);
  const baselineWindow = selectNaiveBaselineWindow(
    scenario.input,
    compiler.aggregate.phaseBeat,
    compiler.aggregate.horizonBeats,
    compiler.aggregate.catchToleranceBeats
  );
  const selectedWindowId = compiler.selectedWindow?.windowId ?? null;
  const baselineWindowId = baselineWindow?.windowId ?? null;
  const sameSelectionAsBaseline =
    selectedWindowId !== null &&
    baselineWindowId !== null &&
    selectedWindowId === baselineWindowId;
  const unsafeBaselineCommitBlocked =
    !compiler.aggregate.committed && baselineWindow !== null;
  const machineOnlyBaselineSelectionBlocked =
    baselineWindow?.machineOnlyCandidate === true &&
    selectedWindowId !== baselineWindowId;

  return {
    scenarioId: scenario.scenarioId,
    label: scenario.label,
    policy: compiler.policy,
    committed: compiler.aggregate.committed,
    authorityReady: compiler.aggregate.observationReady,
    selectedWindowId,
    baselineWindowId,
    baselineSelectedMachineOnly: baselineWindow?.machineOnlyCandidate ?? false,
    unsafeBaselineCommitBlocked,
    machineOnlyBaselineSelectionBlocked,
    sameSelectionAsBaseline,
    candidateCount: compiler.aggregate.candidateCount,
    suppressedMachineOnlyCount: compiler.aggregate.suppressedMachineOnlyCount,
    suppressedOutsideHorizonCount:
      compiler.aggregate.suppressedOutsideHorizonCount,
  };
}

/**
 * Handles the gnosis run Pattern Horizon Benchmark workflow.
 */
export function runPatternHorizonBenchmark(
  scenarios: readonly PatternHorizonBenchmarkScenario[]
): PatternHorizonBenchmarkReport {
  const scenarioResults = scenarios.map(buildScenarioResult);
  const compilerCommits = scenarioResults.filter(
    (scenario) => scenario.committed
  );
  const baselineCommits = scenarioResults.filter(
    (scenario) => scenario.baselineWindowId !== null
  );
  const baselineMachineOnlyCommits = scenarioResults.filter(
    (scenario) => scenario.baselineSelectedMachineOnly
  );
  const hedgeCount = scenarioResults.filter(
    (scenario) => scenario.policy === 'hedge-shadow'
  ).length;
  const escalationCount = scenarioResults.filter(
    (scenario) => scenario.policy === 'escalate-machine-only'
  ).length;
  const sameSelectionCount = scenarioResults.filter(
    (scenario) => scenario.sameSelectionAsBaseline
  ).length;

  return {
    label: 'gnosis-pattern-horizon-benchmark-v1',
    scenarioCount: scenarioResults.length,
    compiler: {
      commitRate: compilerCommits.length / Math.max(1, scenarioResults.length),
      safeStateRate:
        (scenarioResults.length - compilerCommits.length) /
        Math.max(1, scenarioResults.length),
      hedgeShare: hedgeCount / Math.max(1, scenarioResults.length),
      escalationShare: escalationCount / Math.max(1, scenarioResults.length),
      meanCandidateCount: mean(
        scenarioResults.map((scenario) => scenario.candidateCount)
      ),
      meanSuppressedMachineOnlyCount: mean(
        scenarioResults.map((scenario) => scenario.suppressedMachineOnlyCount)
      ),
      meanSuppressedOutsideHorizonCount: mean(
        scenarioResults.map(
          (scenario) => scenario.suppressedOutsideHorizonCount
        )
      ),
    },
    baseline: {
      commitRate: baselineCommits.length / Math.max(1, scenarioResults.length),
      machineOnlyCommitRate:
        baselineMachineOnlyCommits.length / Math.max(1, scenarioResults.length),
    },
    comparisons: {
      unsafeBaselineCommitsBlocked: scenarioResults.filter(
        (scenario) => scenario.unsafeBaselineCommitBlocked
      ).length,
      machineOnlyBaselineSelectionsBlocked: scenarioResults.filter(
        (scenario) => scenario.machineOnlyBaselineSelectionBlocked
      ).length,
      sameSelectionRate:
        sameSelectionCount / Math.max(1, scenarioResults.length),
      divergentSelectionRate:
        (scenarioResults.length - sameSelectionCount) /
        Math.max(1, scenarioResults.length),
    },
    scenarios: scenarioResults,
  };
}
