// @a0n/.ignored_gnosis/src/benchmarks/regime-sweep-benchmark.ts
import { readFile } from 'node:fs/promises';

import { parseGgProgram } from '@a0n/aeon-logic';

import {
  DEFAULT_FOLD_TRAINING_TOPOLOGY_FILES,
  type FoldTrainingStrategy,
} from './fold-training-benchmark';
import {
  DEFAULT_MINI_MOE_ROUTING_TOPOLOGY_FILES,
  type MiniMoeRoutingStrategy,
} from './moe-routing-benchmark';
import {
  bootstrapMeanConfidenceInterval,
  createDeterministicRandom,
  mean,
  shuffleInPlace,
  stdev,
  type ConfidenceInterval,
} from './statistics';
import { lowerUfcsSource } from '../ufcs';

type RegimeSweepStrategy = FoldTrainingStrategy;
type RegimeSweepPointStrategy = 'winner-take-all' | 'early-stop';

export interface RegimeSweepSeedMetrics {
  readonly seed: number;
  readonly trainMeanSquaredError: number;
  readonly evalMeanSquaredError: number;
  readonly exactWithinToleranceFraction: number;
}

export interface RegimeSweepStrategyReport {
  readonly seedMetrics: readonly RegimeSweepSeedMetrics[];
  readonly meanTrainMeanSquaredError: number;
  readonly stdevTrainMeanSquaredError: number;
  readonly meanEvalMeanSquaredError: number;
  readonly evalMeanSquaredErrorCi95: ConfidenceInterval;
  readonly stdevEvalMeanSquaredError: number;
  readonly meanExactWithinToleranceFraction: number;
  readonly exactWithinToleranceFractionCi95: ConfidenceInterval;
  readonly stdevExactWithinToleranceFraction: number;
}

export interface RegimeSweepTopologyMetrics {
  readonly nodeCount: number;
  readonly edgeCount: number;
  readonly structuralBeta1: number;
  readonly unitCount: number;
  readonly parameterCount: number;
}

export interface RegimeSweepPointReport {
  readonly regimeValue: number;
  readonly strategies: Record<RegimeSweepStrategy, RegimeSweepStrategyReport>;
  readonly rankingByEvalMeanSquaredError: readonly RegimeSweepStrategy[];
  readonly bestNonlinearStrategy: RegimeSweepPointStrategy;
  readonly bestNonlinearMeanEvalMeanSquaredError: number;
  readonly linearAdvantageEvalMeanSquaredError: number;
  readonly linearAdvantageExactWithinToleranceFraction: number;
}

export interface RegimeSweepFamilyReport {
  readonly title: string;
  readonly description: string;
  readonly topologyFamily: 'fold-training' | 'moe-routing';
  readonly targetFamily: 'weighted-cancellation' | 'weighted-dual-activation';
  readonly sweepDimension: string;
  readonly topology: RegimeSweepTopologyMetrics;
  readonly regimeValues: readonly number[];
  readonly separationThreshold: number;
  readonly monotonicSlack: number;
  readonly points: readonly RegimeSweepPointReport[];
  readonly firstSeparatedRegimeValue: number | null;
  readonly monotonicAdvantageRecovered: boolean;
  readonly phaseBoundaryRecovered: boolean;
}

export interface GnosisFoldBoundaryRegimeSweepReport {
  readonly label: string;
  readonly config: {
    readonly affineSeedCount: number;
    readonly affineEpochs: number;
    readonly affineLearningRate: number;
    readonly affineExactPredictionTolerance: number;
    readonly affineTrainSampleCount: number;
    readonly affineEvalSampleCount: number;
    readonly routedSeedCount: number;
    readonly routedEpochs: number;
    readonly routedLearningRate: number;
    readonly routedExactPredictionTolerance: number;
    readonly routedTrainSampleCount: number;
    readonly routedEvalSampleCount: number;
    readonly regimeValues: readonly number[];
  };
  readonly affine: RegimeSweepFamilyReport;
  readonly routed: RegimeSweepFamilyReport;
  readonly predictedBoundaryRecovered: boolean;
}

export interface AffineSweepConfig {
  readonly seedCount: number;
  readonly epochs: number;
  readonly learningRate: number;
  readonly exactPredictionTolerance: number;
  readonly trainAxis: readonly number[];
  readonly evalAxis: readonly number[];
  readonly regimeValues: readonly number[];
  readonly separationThreshold: number;
  readonly monotonicSlack: number;
}

export interface RoutedSweepConfig {
  readonly seedCount: number;
  readonly epochs: number;
  readonly learningRate: number;
  readonly exactPredictionTolerance: number;
  readonly trainAxis: readonly number[];
  readonly evalAxis: readonly number[];
  readonly regimeValues: readonly number[];
  readonly separationThreshold: number;
  readonly monotonicSlack: number;
}

interface AffineSample {
  readonly leftInput: number;
  readonly rightInput: number;
  readonly target: number;
}

interface AffineParameters {
  leftWeight: number;
  leftBias: number;
  rightWeight: number;
  rightBias: number;
}

interface AffinePrediction {
  readonly leftValue: number;
  readonly rightValue: number;
  readonly prediction: number;
  readonly selectedBranch: 'left' | 'right';
}

interface RoutedSample {
  readonly leftInput: number;
  readonly rightInput: number;
  readonly target: number;
}

interface ExpertParameters {
  gateWeight: number;
  gateBias: number;
  valueWeight: number;
  valueBias: number;
}

interface ExpertSpec {
  readonly signedProjection: (sample: RoutedSample) => number;
  readonly axisProjection: (sample: RoutedSample) => number;
}

interface ExpertActivation {
  readonly gate: number;
  readonly value: number;
  readonly output: number;
  readonly signedInput: number;
  readonly axisInput: number;
}

interface RoutedPrediction {
  readonly prediction: number;
  readonly selectedExpertIndex: number;
  readonly expertActivations: readonly ExpertActivation[];
}

export interface RegimeSweepRunConfig {
  readonly affine: AffineSweepConfig;
  readonly routed: RoutedSweepConfig;
}

const regimeSweepStrategies = Object.keys(
  DEFAULT_FOLD_TRAINING_TOPOLOGY_FILES
) as RegimeSweepStrategy[];

const expertSpecs: readonly ExpertSpec[] = [
  {
    signedProjection: (sample) => sample.leftInput,
    axisProjection: (sample) => sample.leftInput,
  },
  {
    signedProjection: (sample) => -sample.leftInput,
    axisProjection: (sample) => sample.leftInput,
  },
  {
    signedProjection: (sample) => sample.rightInput,
    axisProjection: (sample) => sample.rightInput,
  },
  {
    signedProjection: (sample) => -sample.rightInput,
    axisProjection: (sample) => sample.rightInput,
  },
] as const;

function sigmoid(value: number): number {
  return 1 / (1 + Math.exp(-value));
}

function countConnectedComponents(
  nodeIds: readonly string[],
  edges: readonly {
    readonly sourceIds: readonly string[];
    readonly targetIds: readonly string[];
  }[]
): number {
  const neighbors = new Map<string, Set<string>>();
  for (const nodeId of nodeIds) {
    neighbors.set(nodeId, new Set<string>());
  }

  for (const edge of edges) {
    for (const sourceId of edge.sourceIds) {
      const sourceNeighbors = neighbors.get(sourceId);
      if (!sourceNeighbors) {
        continue;
      }
      for (const targetId of edge.targetIds) {
        sourceNeighbors.add(targetId);
        neighbors.get(targetId)?.add(sourceId);
      }
    }
  }

  let components = 0;
  const visited = new Set<string>();
  for (const nodeId of nodeIds) {
    if (visited.has(nodeId)) {
      continue;
    }

    components++;
    const frontier = [nodeId];
    while (frontier.length > 0) {
      const current = frontier.pop();
      if (current === undefined || visited.has(current)) {
        continue;
      }
      visited.add(current);
      for (const neighbor of neighbors.get(current) ?? []) {
        if (!visited.has(neighbor)) {
          frontier.push(neighbor);
        }
      }
    }
  }

  return Math.max(1, components);
}

function structuralBeta1(
  nodeIds: readonly string[],
  edges: readonly {
    readonly sourceIds: readonly string[];
    readonly targetIds: readonly string[];
  }[]
): number {
  const expandedEdgeCount = edges.reduce(
    (sum, edge) => sum + edge.sourceIds.length * edge.targetIds.length,
    0
  );
  return Math.max(
    0,
    expandedEdgeCount -
      nodeIds.length +
      countConnectedComponents(nodeIds, edges)
  );
}

async function loadAffineTopologyMetrics(): Promise<RegimeSweepTopologyMetrics> {
  const source = await readFile(
    DEFAULT_FOLD_TRAINING_TOPOLOGY_FILES.linear,
    'utf8'
  );
  const program = parseGgProgram(lowerUfcsSource(source));
  const branchNodes = program.nodes.filter((node) =>
    node.labels.includes('AffineBranch')
  );
  return {
    nodeCount: program.nodes.length,
    edgeCount: program.edges.length,
    structuralBeta1: structuralBeta1(
      program.nodes.map((node) => node.id),
      program.edges.map((edge) => ({
        sourceIds: edge.sourceIds,
        targetIds: edge.targetIds,
      }))
    ),
    unitCount: branchNodes.length,
    parameterCount: branchNodes.length * 2,
  };
}

async function loadRoutedTopologyMetrics(): Promise<RegimeSweepTopologyMetrics> {
  const source = await readFile(
    DEFAULT_MINI_MOE_ROUTING_TOPOLOGY_FILES.linear,
    'utf8'
  );
  const program = parseGgProgram(lowerUfcsSource(source));
  const expertNodes = program.nodes.filter((node) =>
    node.labels.includes('MoEExpert')
  );
  return {
    nodeCount: program.nodes.length,
    edgeCount: program.edges.length,
    structuralBeta1: structuralBeta1(
      program.nodes.map((node) => node.id),
      program.edges.map((edge) => ({
        sourceIds: edge.sourceIds,
        targetIds: edge.targetIds,
      }))
    ),
    unitCount: expertNodes.length,
    parameterCount: expertNodes.reduce((sum, node) => {
      const parameterCount = Number(node.properties.parameters ?? 0);
      return sum + (Number.isFinite(parameterCount) ? parameterCount : 0);
    }, 0),
  };
}

function initializeAffineParameters(seed: number): AffineParameters {
  const random = createDeterministicRandom(seed);
  const span = 0.8;
  const offset = -0.4;
  return {
    leftWeight: offset + random() * span,
    leftBias: offset + random() * span,
    rightWeight: offset + random() * span,
    rightBias: offset + random() * span,
  };
}

function buildAffineSamples(
  axis: readonly number[],
  regimeValue: number
): AffineSample[] {
  return axis.flatMap((leftInput) =>
    axis.map((rightInput) => ({
      leftInput,
      rightInput,
      target: leftInput - regimeValue * rightInput,
    }))
  );
}

function predictAffine(
  strategy: RegimeSweepStrategy,
  parameters: AffineParameters,
  sample: AffineSample
): AffinePrediction {
  const leftValue =
    parameters.leftWeight * sample.leftInput + parameters.leftBias;
  const rightValue =
    parameters.rightWeight * sample.rightInput + parameters.rightBias;

  if (strategy === 'linear') {
    return {
      leftValue,
      rightValue,
      prediction: leftValue + rightValue,
      selectedBranch: 'left',
    };
  }

  if (strategy === 'early-stop') {
    return {
      leftValue,
      rightValue,
      prediction: leftValue,
      selectedBranch: 'left',
    };
  }

  const selectedBranch =
    Math.abs(leftValue) >= Math.abs(rightValue) ? 'left' : 'right';
  return {
    leftValue,
    rightValue,
    prediction: selectedBranch === 'left' ? leftValue : rightValue,
    selectedBranch,
  };
}

function updateAffineParameters(
  strategy: RegimeSweepStrategy,
  parameters: AffineParameters,
  sample: AffineSample,
  learningRate: number
): void {
  const prediction = predictAffine(strategy, parameters, sample);
  const error = prediction.prediction - sample.target;

  if (strategy === 'linear') {
    parameters.leftWeight -= learningRate * error * sample.leftInput;
    parameters.leftBias -= learningRate * error;
    parameters.rightWeight -= learningRate * error * sample.rightInput;
    parameters.rightBias -= learningRate * error;
    return;
  }

  if (prediction.selectedBranch === 'left') {
    parameters.leftWeight -= learningRate * error * sample.leftInput;
    parameters.leftBias -= learningRate * error;
    return;
  }

  parameters.rightWeight -= learningRate * error * sample.rightInput;
  parameters.rightBias -= learningRate * error;
}

function evaluateAffineSamples(
  strategy: RegimeSweepStrategy,
  parameters: AffineParameters,
  samples: readonly AffineSample[],
  tolerance: number
): {
  readonly meanSquaredError: number;
  readonly exactWithinToleranceFraction: number;
} {
  if (samples.length === 0) {
    return {
      meanSquaredError: 0,
      exactWithinToleranceFraction: 0,
    };
  }

  let squaredErrorSum = 0;
  let exactCount = 0;
  for (const sample of samples) {
    const error =
      predictAffine(strategy, parameters, sample).prediction - sample.target;
    squaredErrorSum += error * error;
    if (Math.abs(error) <= tolerance) {
      exactCount++;
    }
  }

  return {
    meanSquaredError: squaredErrorSum / samples.length,
    exactWithinToleranceFraction: exactCount / samples.length,
  };
}

function initializeRoutedParameters(seed: number): ExpertParameters[] {
  const random = createDeterministicRandom(seed);
  return expertSpecs.map(() => ({
    gateWeight: (random() - 0.5) * 1.4,
    gateBias: (random() - 0.5) * 0.5,
    valueWeight: (random() - 0.5) * 0.9,
    valueBias: (random() - 0.5) * 0.3,
  }));
}

function buildRoutedSamples(
  axis: readonly number[],
  regimeValue: number
): RoutedSample[] {
  return axis.flatMap((leftInput) =>
    axis.map((rightInput) => ({
      leftInput,
      rightInput,
      target: Math.abs(leftInput) + regimeValue * Math.abs(rightInput),
    }))
  );
}

function predictRouted(
  strategy: RegimeSweepStrategy,
  parameters: readonly ExpertParameters[],
  sample: RoutedSample
): RoutedPrediction {
  const expertActivations = expertSpecs.map((expert, index) => {
    const parameter = parameters[index];
    if (!parameter) {
      throw new Error(`Missing routed parameters for expert ${index}`);
    }

    const signedInput = expert.signedProjection(sample);
    const axisInput = expert.axisProjection(sample);
    const gate = sigmoid(
      parameter.gateWeight * signedInput + parameter.gateBias
    );
    const value = parameter.valueWeight * axisInput + parameter.valueBias;
    return {
      gate,
      value,
      output: gate * value,
      signedInput,
      axisInput,
    };
  });

  let selectedExpertIndex = 0;
  for (let index = 1; index < expertActivations.length; index++) {
    const selected = expertActivations[selectedExpertIndex];
    const candidate = expertActivations[index];
    if (
      candidate &&
      selected &&
      Math.abs(candidate.output) > Math.abs(selected.output)
    ) {
      selectedExpertIndex = index;
    }
  }

  if (strategy === 'linear') {
    return {
      prediction: expertActivations.reduce(
        (sum, activation) => sum + activation.output,
        0
      ),
      selectedExpertIndex,
      expertActivations,
    };
  }

  if (strategy === 'early-stop') {
    return {
      prediction: expertActivations[0]?.output ?? 0,
      selectedExpertIndex: 0,
      expertActivations,
    };
  }

  return {
    prediction: expertActivations[selectedExpertIndex]?.output ?? 0,
    selectedExpertIndex,
    expertActivations,
  };
}

function updateRoutedParameters(
  strategy: RegimeSweepStrategy,
  parameters: ExpertParameters[],
  sample: RoutedSample,
  learningRate: number
): void {
  const prediction = predictRouted(strategy, parameters, sample);
  const error = prediction.prediction - sample.target;

  for (let index = 0; index < parameters.length; index++) {
    if (
      strategy === 'winner-take-all' &&
      index !== prediction.selectedExpertIndex
    ) {
      continue;
    }
    if (strategy === 'early-stop' && index !== 0) {
      continue;
    }

    const parameter = parameters[index];
    const activation = prediction.expertActivations[index];
    if (!parameter || !activation) {
      continue;
    }

    const gatePrime = activation.gate * (1 - activation.gate);
    parameter.valueWeight -=
      learningRate * error * activation.gate * activation.axisInput;
    parameter.valueBias -= learningRate * error * activation.gate;
    parameter.gateWeight -=
      learningRate *
      error *
      gatePrime *
      activation.signedInput *
      activation.value;
    parameter.gateBias -= learningRate * error * gatePrime * activation.value;
  }
}

function evaluateRoutedSamples(
  strategy: RegimeSweepStrategy,
  parameters: readonly ExpertParameters[],
  samples: readonly RoutedSample[],
  tolerance: number
): {
  readonly meanSquaredError: number;
  readonly exactWithinToleranceFraction: number;
} {
  if (samples.length === 0) {
    return {
      meanSquaredError: 0,
      exactWithinToleranceFraction: 0,
    };
  }

  let squaredErrorSum = 0;
  let exactCount = 0;
  for (const sample of samples) {
    const error =
      predictRouted(strategy, parameters, sample).prediction - sample.target;
    squaredErrorSum += error * error;
    if (Math.abs(error) <= tolerance) {
      exactCount++;
    }
  }

  return {
    meanSquaredError: squaredErrorSum / samples.length,
    exactWithinToleranceFraction: exactCount / samples.length,
  };
}

function summarizeSeedMetrics(
  metrics: readonly RegimeSweepSeedMetrics[],
  ciSeed: number
): RegimeSweepStrategyReport {
  const evalMeanSquaredErrors = metrics.map(
    (entry) => entry.evalMeanSquaredError
  );
  const exactFractions = metrics.map(
    (entry) => entry.exactWithinToleranceFraction
  );

  return {
    seedMetrics: metrics,
    meanTrainMeanSquaredError: mean(
      metrics.map((entry) => entry.trainMeanSquaredError)
    ),
    stdevTrainMeanSquaredError: stdev(
      metrics.map((entry) => entry.trainMeanSquaredError)
    ),
    meanEvalMeanSquaredError: mean(evalMeanSquaredErrors),
    evalMeanSquaredErrorCi95: bootstrapMeanConfidenceInterval(
      evalMeanSquaredErrors,
      {
        seed: ciSeed,
      }
    ),
    stdevEvalMeanSquaredError: stdev(evalMeanSquaredErrors),
    meanExactWithinToleranceFraction: mean(exactFractions),
    exactWithinToleranceFractionCi95: bootstrapMeanConfidenceInterval(
      exactFractions,
      {
        seed: ciSeed ^ 0x00abcdef,
      }
    ),
    stdevExactWithinToleranceFraction: stdev(exactFractions),
  };
}

function rankStrategiesByEvalMse(
  strategies: Record<RegimeSweepStrategy, RegimeSweepStrategyReport>
): RegimeSweepStrategy[] {
  return [...regimeSweepStrategies].sort(
    (left, right) =>
      strategies[left].meanEvalMeanSquaredError -
      strategies[right].meanEvalMeanSquaredError
  );
}

function bestNonlinearStrategy(
  strategies: Record<RegimeSweepStrategy, RegimeSweepStrategyReport>
): RegimeSweepPointStrategy {
  return strategies['winner-take-all'].meanEvalMeanSquaredError <=
    strategies['early-stop'].meanEvalMeanSquaredError
    ? 'winner-take-all'
    : 'early-stop';
}

async function runAffinePoint(
  config: AffineSweepConfig,
  regimeValue: number
): Promise<RegimeSweepPointReport> {
  const trainSamples = buildAffineSamples(config.trainAxis, regimeValue);
  const evalSamples = buildAffineSamples(config.evalAxis, regimeValue);
  const strategies = await Promise.all(
    regimeSweepStrategies.map(async (strategy, strategyIndex) => {
      const trainedMetrics: RegimeSweepSeedMetrics[] = [];
      for (let seed = 1; seed <= config.seedCount; seed++) {
        const random = createDeterministicRandom(
          0x9100 + strategyIndex * 101 + seed
        );
        const parameters = initializeAffineParameters(
          seed + strategyIndex * 100
        );
        const shuffledSamples = [...trainSamples];
        for (let epoch = 0; epoch < config.epochs; epoch++) {
          shuffleInPlace(shuffledSamples, random);
          for (const sample of shuffledSamples) {
            updateAffineParameters(
              strategy,
              parameters,
              sample,
              config.learningRate
            );
          }
        }

        const trainMetrics = evaluateAffineSamples(
          strategy,
          parameters,
          trainSamples,
          config.exactPredictionTolerance
        );
        const evalMetrics = evaluateAffineSamples(
          strategy,
          parameters,
          evalSamples,
          config.exactPredictionTolerance
        );
        trainedMetrics.push({
          seed,
          trainMeanSquaredError: trainMetrics.meanSquaredError,
          evalMeanSquaredError: evalMetrics.meanSquaredError,
          exactWithinToleranceFraction:
            evalMetrics.exactWithinToleranceFraction,
        });
      }

      return [
        strategy,
        summarizeSeedMetrics(trainedMetrics, 0x410000 + strategyIndex * 257),
      ] as const;
    })
  );

  const strategyReports = Object.fromEntries(strategies) as Record<
    RegimeSweepStrategy,
    RegimeSweepStrategyReport
  >;
  const nonlinearStrategy = bestNonlinearStrategy(strategyReports);
  return {
    regimeValue,
    strategies: strategyReports,
    rankingByEvalMeanSquaredError: rankStrategiesByEvalMse(strategyReports),
    bestNonlinearStrategy: nonlinearStrategy,
    bestNonlinearMeanEvalMeanSquaredError:
      strategyReports[nonlinearStrategy].meanEvalMeanSquaredError,
    linearAdvantageEvalMeanSquaredError:
      strategyReports[nonlinearStrategy].meanEvalMeanSquaredError -
      strategyReports.linear.meanEvalMeanSquaredError,
    linearAdvantageExactWithinToleranceFraction:
      strategyReports.linear.meanExactWithinToleranceFraction -
      strategyReports[nonlinearStrategy].meanExactWithinToleranceFraction,
  };
}

async function runRoutedPoint(
  config: RoutedSweepConfig,
  regimeValue: number
): Promise<RegimeSweepPointReport> {
  const trainSamples = buildRoutedSamples(config.trainAxis, regimeValue);
  const evalSamples = buildRoutedSamples(config.evalAxis, regimeValue);
  const strategies = await Promise.all(
    regimeSweepStrategies.map(async (strategy, strategyIndex) => {
      const trainedMetrics: RegimeSweepSeedMetrics[] = [];
      for (let seed = 1; seed <= config.seedCount; seed++) {
        const random = createDeterministicRandom(
          0xa200 + strategyIndex * 149 + seed
        );
        const parameters = initializeRoutedParameters(
          seed + strategyIndex * 100
        );
        const shuffledSamples = [...trainSamples];
        for (let epoch = 0; epoch < config.epochs; epoch++) {
          shuffleInPlace(shuffledSamples, random);
          for (const sample of shuffledSamples) {
            updateRoutedParameters(
              strategy,
              parameters,
              sample,
              config.learningRate
            );
          }
        }

        const trainMetrics = evaluateRoutedSamples(
          strategy,
          parameters,
          trainSamples,
          config.exactPredictionTolerance
        );
        const evalMetrics = evaluateRoutedSamples(
          strategy,
          parameters,
          evalSamples,
          config.exactPredictionTolerance
        );
        trainedMetrics.push({
          seed,
          trainMeanSquaredError: trainMetrics.meanSquaredError,
          evalMeanSquaredError: evalMetrics.meanSquaredError,
          exactWithinToleranceFraction:
            evalMetrics.exactWithinToleranceFraction,
        });
      }

      return [
        strategy,
        summarizeSeedMetrics(trainedMetrics, 0x520000 + strategyIndex * 257),
      ] as const;
    })
  );

  const strategyReports = Object.fromEntries(strategies) as Record<
    RegimeSweepStrategy,
    RegimeSweepStrategyReport
  >;
  const nonlinearStrategy = bestNonlinearStrategy(strategyReports);
  return {
    regimeValue,
    strategies: strategyReports,
    rankingByEvalMeanSquaredError: rankStrategiesByEvalMse(strategyReports),
    bestNonlinearStrategy: nonlinearStrategy,
    bestNonlinearMeanEvalMeanSquaredError:
      strategyReports[nonlinearStrategy].meanEvalMeanSquaredError,
    linearAdvantageEvalMeanSquaredError:
      strategyReports[nonlinearStrategy].meanEvalMeanSquaredError -
      strategyReports.linear.meanEvalMeanSquaredError,
    linearAdvantageExactWithinToleranceFraction:
      strategyReports.linear.meanExactWithinToleranceFraction -
      strategyReports[nonlinearStrategy].meanExactWithinToleranceFraction,
  };
}

function findFirstSeparatedRegimeValue(
  points: readonly RegimeSweepPointReport[],
  threshold: number
): number | null {
  const firstSeparated = points.find(
    (point) => point.linearAdvantageEvalMeanSquaredError > threshold
  );
  return firstSeparated?.regimeValue ?? null;
}

function checkMonotonicAdvantage(
  points: readonly RegimeSweepPointReport[],
  slack: number
): boolean {
  for (let index = 1; index < points.length; index++) {
    const previous = points[index - 1];
    const current = points[index];
    if (
      previous &&
      current &&
      current.linearAdvantageEvalMeanSquaredError + slack <
        previous.linearAdvantageEvalMeanSquaredError
    ) {
      return false;
    }
  }
  return true;
}

/**
 * Handles the gnosis make Default Regime Sweep Config workflow.
 */
export function makeDefaultRegimeSweepConfig(): RegimeSweepRunConfig {
  const regimeValues = [0, 0.25, 0.5, 0.75, 1] as const;
  return {
    affine: {
      seedCount: 6,
      epochs: 420,
      learningRate: 0.035,
      exactPredictionTolerance: 0.05,
      trainAxis: [-1.5, -1.0, -0.5, 0, 0.5, 1.0, 1.5],
      evalAxis: [-1.25, -0.75, -0.25, 0.25, 0.75, 1.25],
      regimeValues,
      separationThreshold: 0.08,
      monotonicSlack: 0.03,
    },
    routed: {
      seedCount: 8,
      epochs: 720,
      learningRate: 0.025,
      exactPredictionTolerance: 0.08,
      trainAxis: [0.2, 0.4, 0.6, 0.8, 1.0, 1.2, 1.4],
      evalAxis: [0.25, 0.45, 0.65, 0.85, 1.05, 1.25],
      regimeValues,
      separationThreshold: 0.05,
      monotonicSlack: 0.03,
    },
  };
}

/**
 * Builds the Gnosis Fold Boundary Regime Sweep Report.
 */
export async function buildGnosisFoldBoundaryRegimeSweepReport(
  config: RegimeSweepRunConfig,
  label = 'gnosis-fold-boundary-regime-sweep-v1'
): Promise<GnosisFoldBoundaryRegimeSweepReport> {
  const [affineTopology, routedTopology] = await Promise.all([
    loadAffineTopologyMetrics(),
    loadRoutedTopologyMetrics(),
  ]);

  const affinePoints = await Promise.all(
    config.affine.regimeValues.map((regimeValue) =>
      runAffinePoint(config.affine, regimeValue)
    )
  );
  const routedPoints = await Promise.all(
    config.routed.regimeValues.map((regimeValue) =>
      runRoutedPoint(config.routed, regimeValue)
    )
  );

  const affineFirstSeparated = findFirstSeparatedRegimeValue(
    affinePoints,
    config.affine.separationThreshold
  );
  const routedFirstSeparated = findFirstSeparatedRegimeValue(
    routedPoints,
    config.routed.separationThreshold
  );
  const affinePhaseBoundaryRecovered =
    affinePoints[0] !== undefined &&
    affinePoints[0].linearAdvantageEvalMeanSquaredError <=
      config.affine.separationThreshold &&
    affinePoints[affinePoints.length - 1] !== undefined &&
    (affinePoints[affinePoints.length - 1]
      ?.linearAdvantageEvalMeanSquaredError ?? 0) >
      config.affine.separationThreshold &&
    affineFirstSeparated !== null;
  const routedPhaseBoundaryRecovered =
    routedPoints[0] !== undefined &&
    routedPoints[0].linearAdvantageEvalMeanSquaredError <=
      config.routed.separationThreshold &&
    routedPoints[routedPoints.length - 1] !== undefined &&
    (routedPoints[routedPoints.length - 1]
      ?.linearAdvantageEvalMeanSquaredError ?? 0) >
      config.routed.separationThreshold &&
    routedFirstSeparated !== null;

  return {
    label,
    config: {
      affineSeedCount: config.affine.seedCount,
      affineEpochs: config.affine.epochs,
      affineLearningRate: config.affine.learningRate,
      affineExactPredictionTolerance: config.affine.exactPredictionTolerance,
      affineTrainSampleCount: config.affine.trainAxis.length ** 2,
      affineEvalSampleCount: config.affine.evalAxis.length ** 2,
      routedSeedCount: config.routed.seedCount,
      routedEpochs: config.routed.epochs,
      routedLearningRate: config.routed.learningRate,
      routedExactPredictionTolerance: config.routed.exactPredictionTolerance,
      routedTrainSampleCount: config.routed.trainAxis.length ** 2,
      routedEvalSampleCount: config.routed.evalAxis.length ** 2,
      regimeValues: config.affine.regimeValues,
    },
    affine: {
      title: 'Affine cancellation regime sweep',
      description:
        'Reuses the two-branch affine topology while increasing the right-branch cancellation weight from one-path parity to full left-minus-right cancellation.',
      topologyFamily: 'fold-training',
      targetFamily: 'weighted-cancellation',
      sweepDimension: 'cancellation weight on the right branch',
      topology: affineTopology,
      regimeValues: config.affine.regimeValues,
      separationThreshold: config.affine.separationThreshold,
      monotonicSlack: config.affine.monotonicSlack,
      points: affinePoints,
      firstSeparatedRegimeValue: affineFirstSeparated,
      monotonicAdvantageRecovered: checkMonotonicAdvantage(
        affinePoints,
        config.affine.monotonicSlack
      ),
      phaseBoundaryRecovered: affinePhaseBoundaryRecovered,
    },
    routed: {
      title: 'Routed dual-activation regime sweep',
      description:
        'Reuses the four-expert routed topology while increasing the y-axis contribution weight from single-expert parity to full dual-axis dependence.',
      topologyFamily: 'moe-routing',
      targetFamily: 'weighted-dual-activation',
      sweepDimension: 'dual-activation weight on the y-axis contribution',
      topology: routedTopology,
      regimeValues: config.routed.regimeValues,
      separationThreshold: config.routed.separationThreshold,
      monotonicSlack: config.routed.monotonicSlack,
      points: routedPoints,
      firstSeparatedRegimeValue: routedFirstSeparated,
      monotonicAdvantageRecovered: checkMonotonicAdvantage(
        routedPoints,
        config.routed.monotonicSlack
      ),
      phaseBoundaryRecovered: routedPhaseBoundaryRecovered,
    },
    predictedBoundaryRecovered:
      affinePhaseBoundaryRecovered &&
      routedPhaseBoundaryRecovered &&
      checkMonotonicAdvantage(affinePoints, config.affine.monotonicSlack) &&
      checkMonotonicAdvantage(routedPoints, config.routed.monotonicSlack),
  };
}

/**
 * Handles the gnosis run Gnosis Fold Boundary Regime Sweep workflow.
 */
export async function runGnosisFoldBoundaryRegimeSweep(): Promise<GnosisFoldBoundaryRegimeSweepReport> {
  return buildGnosisFoldBoundaryRegimeSweepReport(
    makeDefaultRegimeSweepConfig()
  );
}

function renderFamilyMarkdown(family: RegimeSweepFamilyReport): string {
  const lines: string[] = [];
  lines.push(`## ${family.title}`);
  lines.push('');
  lines.push(`- Description: ${family.description}`);
  lines.push(`- Topology family: \`${family.topologyFamily}\``);
  lines.push(`- Target family: \`${family.targetFamily}\``);
  lines.push(`- Sweep dimension: ${family.sweepDimension}`);
  lines.push(
    `- Shared topology: β₁=${family.topology.structuralBeta1}, units=${family.topology.unitCount}, parameters=${family.topology.parameterCount}`
  );
  lines.push(
    `- First separated regime value: \`${
      family.firstSeparatedRegimeValue === null
        ? 'none'
        : family.firstSeparatedRegimeValue.toFixed(2)
    }\``
  );
  lines.push(
    `- Phase boundary recovered: \`${
      family.phaseBoundaryRecovered ? 'yes' : 'no'
    }\``
  );
  lines.push(
    `- Monotonic advantage recovered: \`${
      family.monotonicAdvantageRecovered ? 'yes' : 'no'
    }\``
  );
  lines.push('');
  lines.push(
    '| Regime | Linear eval MSE | Best nonlinear | Best nonlinear eval MSE | Linear advantage | Linear exact | Best nonlinear exact | Ranking |'
  );
  lines.push('| ---: | ---: | --- | ---: | ---: | ---: | ---: | --- |');
  for (const point of family.points) {
    const bestNonlinear = point.bestNonlinearStrategy;
    lines.push(
      `| ${point.regimeValue.toFixed(
        2
      )} | ${point.strategies.linear.meanEvalMeanSquaredError.toFixed(
        3
      )} | \`${bestNonlinear}\` | ${point.bestNonlinearMeanEvalMeanSquaredError.toFixed(
        3
      )} | ${point.linearAdvantageEvalMeanSquaredError.toFixed(
        3
      )} | ${point.strategies.linear.meanExactWithinToleranceFraction.toFixed(
        3
      )} | ${point.strategies[
        bestNonlinear
      ].meanExactWithinToleranceFraction.toFixed(
        3
      )} | ${point.rankingByEvalMeanSquaredError
        .map((strategy) => `\`${strategy}\``)
        .join(' < ')} |`
    );
  }
  lines.push('');
  return lines.join('\n');
}

/**
 * Handles the gnosis render Gnosis Fold Boundary Regime Sweep Markdown workflow.
 */
export function renderGnosisFoldBoundaryRegimeSweepMarkdown(
  report: GnosisFoldBoundaryRegimeSweepReport
): string {
  const lines: string[] = [];
  lines.push('# Gnosis Fold Boundary Regime Sweep');
  lines.push('');
  lines.push(`- Label: \`${report.label}\``);
  lines.push(
    `- Predicted boundary recovered: \`${
      report.predictedBoundaryRecovered ? 'yes' : 'no'
    }\``
  );
  lines.push(
    `- Affine seeds/epochs: \`${report.config.affineSeedCount}\` / \`${report.config.affineEpochs}\``
  );
  lines.push(
    `- Routed seeds/epochs: \`${report.config.routedSeedCount}\` / \`${report.config.routedEpochs}\``
  );
  lines.push(
    `- Regime values: ${report.config.regimeValues
      .map((value) => `\`${value.toFixed(2)}\``)
      .join(', ')}`
  );
  lines.push('');
  lines.push(renderFamilyMarkdown(report.affine));
  lines.push('');
  lines.push(renderFamilyMarkdown(report.routed));
  lines.push('');
  lines.push(
    'Interpretation: both sweeps start in the one-path parity regime, then move into task families that increasingly require additive recombination. The reported linear advantage quantifies where the nonlinear selection rules stop matching the linear baseline and how sharply that separation grows.'
  );
  lines.push('');
  return `${lines.join('\n')}\n`;
}
