// @a0n/.ignored_gnosis/src/benchmarks/statistics.ts
export interface ConfidenceInterval {
  readonly low: number;
  readonly high: number;
}

interface BootstrapOptions {
  readonly confidenceLevel?: number;
  readonly resampleCount?: number;
  readonly seed?: number;
}

/**
 * Handles the gnosis mean workflow.
 */
export function mean(values: readonly number[]): number {
  if (values.length === 0) {
    return 0;
  }

  return values.reduce((sum, value) => sum + value, 0) / values.length;
}

/**
 * Handles the gnosis stdev workflow.
 */
export function stdev(values: readonly number[]): number {
  if (values.length <= 1) {
    return 0;
  }

  const center = mean(values);
  const variance = mean(
    values.map((value) => {
      const delta = value - center;
      return delta * delta;
    })
  );

  return Math.sqrt(variance);
}

/**
 * Creates the Deterministic Random.
 */
export function createDeterministicRandom(seed: number): () => number {
  let state = seed >>> 0;
  if (state === 0) {
    state = 0x6d2b79f5;
  }

  return () => {
    state = (1664525 * state + 1013904223) >>> 0;
    return state / 0x100000000;
  };
}

export function shuffleInPlace<T>(values: T[], random: () => number): void {
  for (let index = values.length - 1; index > 0; index--) {
    const swapIndex = Math.floor(random() * (index + 1));
    const currentValue = values[index];
    values[index] = values[swapIndex] as T;
    values[swapIndex] = currentValue as T;
  }
}

function percentile(sortedValues: readonly number[], fraction: number): number {
  if (sortedValues.length === 0) {
    return 0;
  }

  if (sortedValues.length === 1) {
    return sortedValues[0] ?? 0;
  }

  const clampedFraction = Math.min(1, Math.max(0, fraction));
  const scaledIndex = clampedFraction * (sortedValues.length - 1);
  const lowerIndex = Math.floor(scaledIndex);
  const upperIndex = Math.ceil(scaledIndex);
  const lowerValue = sortedValues[lowerIndex] ?? sortedValues[0] ?? 0;
  const upperValue =
    sortedValues[upperIndex] ?? sortedValues[sortedValues.length - 1] ?? 0;

  if (lowerIndex === upperIndex) {
    return lowerValue;
  }

  const weight = scaledIndex - lowerIndex;
  return lowerValue * (1 - weight) + upperValue * weight;
}

/**
 * Handles the gnosis bootstrap Mean Confidence Interval workflow.
 */
export function bootstrapMeanConfidenceInterval(
  values: readonly number[],
  options: BootstrapOptions = {}
): ConfidenceInterval {
  if (values.length === 0) {
    return { low: 0, high: 0 };
  }

  if (values.length === 1) {
    const onlyValue = values[0] ?? 0;
    return { low: onlyValue, high: onlyValue };
  }

  const confidenceLevel = options.confidenceLevel ?? 0.95;
  const resampleCount = options.resampleCount ?? 2048;
  const random = createDeterministicRandom(options.seed ?? 0x31415926);
  const sampleMeans = Array.from({ length: resampleCount }, () => {
    const resample = Array.from({ length: values.length }, () => {
      const index = Math.floor(random() * values.length);
      return values[index] ?? 0;
    });
    return mean(resample);
  }).sort((left, right) => left - right);

  const tailMass = (1 - confidenceLevel) / 2;
  return {
    low: percentile(sampleMeans, tailMass),
    high: percentile(sampleMeans, 1 - tailMass),
  };
}
