// @a0n/.ignored_gnosis/src/benchmarks/unspeech-utf16-brotli-shootout.ts
#!/usr/bin/env bun
/**
 * Unspeech UTF-16 vs Brotli shootout.
 *
 * Compares teleported UTF-16 strategies:
 * - direct
 * - direct-reordered
 * - residual (when baseline is provided)
 * - structural-unordered
 *
 * against Brotli over the same UTF-16LE byte stream.
 */

import { brotliCompressSync, constants as zlibConstants } from 'node:zlib';
import { pathToFileURL } from 'node:url';
import {
  buildUtf16Lexicon,
  buildUtf16StructuralPacket,
  teleportUtf16ThroughLexicon,
  type UnicodeStructuralPacket,
  type UnicodeTeleportPacket,
} from '../unspeech-unicode-teleport';

interface Scenario {
  name: string;
  text: string;
  baselineText?: string;
}

interface StrategyStat {
  strategy: string;
  bytes: number;
  preservesOrder: boolean;
}

interface ScenarioReport {
  name: string;
  rawUtf16Bytes: number;
  brotliUtf16Bytes: number;
  strategies: StrategyStat[];
  bestLossless: StrategyStat;
  bestAny: StrategyStat;
}

function utf16leBytes(text: string): Uint8Array {
  return new Uint8Array(Buffer.from(text, 'utf16le'));
}

function brotliUtf16Bytes(text: string): number {
  const compressed = brotliCompressSync(utf16leBytes(text), {
    params: {
      [zlibConstants.BROTLI_PARAM_QUALITY]: 11,
      [zlibConstants.BROTLI_PARAM_MODE]: zlibConstants.BROTLI_MODE_TEXT,
    },
  });
  return compressed.length;
}

function compareStats(a: StrategyStat, b: StrategyStat): number {
  if (a.bytes !== b.bytes) return a.bytes - b.bytes;
  return a.strategy.localeCompare(b.strategy);
}

function varUintByteLength(value: number): number {
  if (!Number.isInteger(value) || value < 0) {
    throw new RangeError(`varuint expects non-negative integer, received ${value}`);
  }
  let bytes = 1;
  let remainder = Math.floor(value / 128);
  while (remainder > 0) {
    bytes += 1;
    remainder = Math.floor(remainder / 128);
  }
  return bytes;
}

function transportBytesForTeleportPacket(
  packet: UnicodeTeleportPacket | UnicodeStructuralPacket
): number {
  if (packet.strategy === 'direct') {
    return (
      1 +
      varUintByteLength(packet.length) +
      varUintByteLength(packet.firstIndex) +
      varUintByteLength(packet.signPacket.payloadLength) +
      packet.signPacket.payloadLength +
      varUintByteLength(packet.magnitudeBytes.length) +
      packet.magnitudeBytes.length
    );
  }
  if (packet.strategy === 'direct-reordered') {
    return (
      1 +
      varUintByteLength(packet.length) +
      varUintByteLength(packet.firstIndex) +
      varUintByteLength(packet.signPacket.payloadLength) +
      packet.signPacket.payloadLength +
      varUintByteLength(packet.magnitudeBytes.length) +
      packet.magnitudeBytes.length +
      varUintByteLength(packet.permutationBytes.length) +
      packet.permutationBytes.length
    );
  }
  if (packet.strategy === 'residual') {
    return (
      1 +
      varUintByteLength(packet.baselineLength) +
      varUintByteLength(packet.targetLength) +
      varUintByteLength(packet.patchCount) +
      varUintByteLength(packet.patchBytes.length) +
      packet.patchBytes.length
    );
  }
  return (
    1 +
    varUintByteLength(packet.totalSymbols) +
    varUintByteLength(packet.nonZeroBuckets) +
    varUintByteLength(packet.bucketBytes.length) +
    packet.bucketBytes.length
  );
}

async function runScenario(scenario: Scenario): Promise<ScenarioReport> {
  const lexicon = buildUtf16Lexicon(
    scenario.baselineText
      ? [scenario.baselineText, scenario.text]
      : [scenario.text]
  );

  const direct = await teleportUtf16ThroughLexicon(scenario.text, lexicon, {
    mode: 'direct',
    tritonCompression: {
      motifSize: 5,
      useBitwiseFold: false,
      preferSmallerPayload: true,
    },
  });

  const directReordered = await teleportUtf16ThroughLexicon(
    scenario.text,
    lexicon,
    {
      mode: 'direct-reordered',
      tritonCompression: {
        motifSize: 5,
        useBitwiseFold: false,
        preferSmallerPayload: true,
      },
    }
  );

  const structural = buildUtf16StructuralPacket(scenario.text, lexicon);

  const strategies: StrategyStat[] = [
    {
      strategy: direct.strategy,
      bytes: transportBytesForTeleportPacket(direct),
      preservesOrder: true,
    },
    {
      strategy: directReordered.strategy,
      bytes: transportBytesForTeleportPacket(directReordered),
      preservesOrder: true,
    },
    {
      strategy: structural.strategy,
      bytes: transportBytesForTeleportPacket(structural),
      preservesOrder: false,
    },
  ];

  if (scenario.baselineText) {
    const residual = await teleportUtf16ThroughLexicon(scenario.text, lexicon, {
      mode: 'residual',
      baselineText: scenario.baselineText,
      tritonCompression: {
        motifSize: 5,
        useBitwiseFold: false,
        preferSmallerPayload: true,
      },
    });

    strategies.push({
      strategy: residual.strategy,
      bytes: transportBytesForTeleportPacket(residual),
      preservesOrder: true,
    });
  }

  const ordered = [...strategies].sort(compareStats);
  const bestLossless = ordered.find((entry) => entry.preservesOrder);
  if (!bestLossless) {
    throw new Error(`scenario ${scenario.name} has no lossless strategy`);
  }

  return {
    name: scenario.name,
    rawUtf16Bytes: utf16leBytes(scenario.text).length,
    brotliUtf16Bytes: brotliUtf16Bytes(scenario.text),
    strategies: ordered,
    bestLossless,
    bestAny: ordered[0]!,
  };
}

function pctSavings(rawBytes: number, compressedBytes: number): string {
  if (rawBytes <= 0) return '0.0%';
  const savings = (1 - compressedBytes / rawBytes) * 100;
  return `${savings.toFixed(1)}%`;
}

function printReport(report: ScenarioReport): void {
  console.log(`scenario: ${report.name}`);
  console.log(`  raw utf16 bytes  : ${report.rawUtf16Bytes}`);
  console.log(
    `  brotli utf16    : ${report.brotliUtf16Bytes} (${pctSavings(
      report.rawUtf16Bytes,
      report.brotliUtf16Bytes
    )} saved)`
  );
  for (const strategy of report.strategies) {
    const label = strategy.preservesOrder
      ? strategy.strategy
      : `${strategy.strategy} (unordered)`;
    console.log(
      `  ${label.padEnd(30)} ${String(strategy.bytes).padStart(8)} (${pctSavings(
        report.rawUtf16Bytes,
        strategy.bytes
      )} saved)`
    );
  }
  const brotliBeatsLossless = report.brotliUtf16Bytes < report.bestLossless.bytes;
  const brotliBeatsAny = report.brotliUtf16Bytes < report.bestAny.bytes;
  console.log(
    `  winner lossless : ${report.bestLossless.strategy} (${report.bestLossless.bytes})`
  );
  console.log(`  winner any      : ${report.bestAny.strategy} (${report.bestAny.bytes})`);
  console.log(`  brotli < lossless winner: ${brotliBeatsLossless}`);
  console.log(`  brotli < any winner     : ${brotliBeatsAny}`);
  console.log('');
}

/**
 * Runs the gnosis command-line workflow.
 */
export async function main(): Promise<void> {
  const scenarios: Scenario[] = [
    {
      name: 'baseline-close-edit',
      baselineText: 'alpha bravo charlie '.repeat(40),
      text: 'alpha bravo charlie '.repeat(40) + 'delta',
    },
    {
      name: 'order-irrelevant-heavy-repetition',
      text: 'ab'.repeat(4500) + String.fromCodePoint(0x1f642).repeat(2500),
    },
    {
      name: 'english-prose-medium',
      text:
        'Compression is representation under constraints. '.repeat(120) +
        'Unspeech moves signal into shared structure.',
    },
    {
      name: 'high-entropy-unicode-mix',
      text: (() => {
        const alphabet = [
          ...'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789',
          String.fromCodePoint(0x1f642),
          String.fromCodePoint(0x1f680),
          '∑',
          'λ',
          '中',
          '文',
        ];
        let out = '';
        for (let i = 0; i < 6000; i++) {
          out += alphabet[(Math.imul(i, 1103515245) + 12345) % alphabet.length]!;
        }
        return out;
      })(),
    },
  ];

  for (const scenario of scenarios) {
    const report = await runScenario(scenario);
    printReport(report);
  }
}

const invokedPath = process.argv[1];
if (invokedPath && import.meta.url === pathToFileURL(invokedPath).href) {
  main().catch((error: unknown) => {
    console.error(error);
    process.exit(1);
  });
}
