// @a0n/.ignored_gnosis/src/benchmarks/worker-runtime-benchmark.ts
import { BettyCompiler } from '../betty/compiler';
import { GnosisEngine } from '../runtime/engine';
import { GnosisRegistry } from '../runtime/registry';
import { loadGnosisWorkerRuntime } from '../runtime/worker-runtime';
import type { SyncHandler } from '../runtime/compiled-topology';

export interface WorkerRuntimeBenchmarkResult {
  readonly iterations: number;
  readonly engineMs: number;
  readonly workerAsyncMs: number;
  readonly workerSyncMs: number;
  readonly asyncSpeedup: number;
  readonly syncSpeedup: number;
  readonly syncVsAsyncSpeedup: number;
  readonly nativeEdgesProcessed: number;
}

function makeRegistry(): GnosisRegistry {
  const registry = new GnosisRegistry();
  registry.register('Step', async (payload, props) => {
    const delta = Number.parseInt(props.delta ?? '1', 10);
    return (payload as number) + delta;
  });
  registry.register('Forker', async (payload) => [payload, payload]);
  registry.register('Left', async (payload) => (payload as number) + 2);
  registry.register('Right', async (payload) => (payload as number) + 5);
  registry.register('Sink', async (payload) => payload);
  return registry;
}

function makeSyncHandlers(): Map<string, SyncHandler> {
  return new Map<string, SyncHandler>([
    [
      'Step',
      (payload, props) => {
        const delta = Number.parseInt(props.delta ?? '1', 10);
        return (payload as number) + delta;
      },
    ],
    ['Forker', (payload) => [payload, payload]],
    ['Left', (payload) => (payload as number) + 2],
    ['Right', (payload) => (payload as number) + 5],
    ['Sink', (payload) => payload],
  ]);
}

/**
 * Handles the gnosis run Worker Runtime Benchmark workflow.
 */
export async function runWorkerRuntimeBenchmark(
  iterations = 5_000
): Promise<WorkerRuntimeBenchmarkResult> {
  const compiler = new BettyCompiler();
  const registry = makeRegistry();
  const { ast } = compiler.parse(`
    (seed:Forker)-[:FORK]->(left:Left|right:Right)
    (left|right)-[:FOLD]->(sink:Sink)
  `);
  if (!ast) {
    throw new Error('Benchmark topology did not compile.');
  }

  const engine = new GnosisEngine(registry);
  const { runtime: asyncRuntime, nativeSnapshot } =
    await loadGnosisWorkerRuntime({
      ast,
      registry,
      fingerprint: 'worker-runtime-async-benchmark',
    });
  const { runtime: syncRuntime } = await loadGnosisWorkerRuntime({
    ast,
    registry,
    fingerprint: 'worker-runtime-sync-benchmark',
    nativeControlPlane: false,
    syncHandlerOverrides: makeSyncHandlers(),
  });

  await engine.executeWithResult(ast, 0);
  await asyncRuntime.execute(0);
  await syncRuntime.execute(0);

  const engineStart = performance.now();
  for (let i = 0; i < iterations; i++) {
    await engine.executeWithResult(ast, i);
  }
  const engineMs = performance.now() - engineStart;

  const workerAsyncStart = performance.now();
  for (let i = 0; i < iterations; i++) {
    await asyncRuntime.execute(i);
  }
  const workerAsyncMs = performance.now() - workerAsyncStart;

  const workerSyncStart = performance.now();
  for (let i = 0; i < iterations; i++) {
    await syncRuntime.execute(i);
  }
  const workerSyncMs = performance.now() - workerSyncStart;

  return {
    iterations,
    engineMs,
    workerAsyncMs,
    workerSyncMs,
    asyncSpeedup: engineMs / workerAsyncMs,
    syncSpeedup: engineMs / workerSyncMs,
    syncVsAsyncSpeedup: workerAsyncMs / workerSyncMs,
    nativeEdgesProcessed: nativeSnapshot?.edgesProcessed ?? 0,
  };
}

if (import.meta.main) {
  const iterations = Number.parseInt(process.argv[2] ?? '5000', 10);
  const result = await runWorkerRuntimeBenchmark(iterations);
  console.log(
    JSON.stringify(
      {
        iterations: result.iterations,
        engineMs: Number(result.engineMs.toFixed(3)),
        workerAsyncMs: Number(result.workerAsyncMs.toFixed(3)),
        workerSyncMs: Number(result.workerSyncMs.toFixed(3)),
        asyncSpeedup: Number(result.asyncSpeedup.toFixed(2)),
        syncSpeedup: Number(result.syncSpeedup.toFixed(2)),
        syncVsAsyncSpeedup: Number(result.syncVsAsyncSpeedup.toFixed(2)),
        nativeEdgesProcessed: result.nativeEdgesProcessed,
      },
      null,
      2
    )
  );
}
