// @a0n/.ignored_gnosis/src/testing/index.ts
export {
  GGTestBuilder,
  ggAssert,
  ggQuickCheck,
  ggTest,
} from '../gg-test-harness';
export {
  discoverTestFiles as discoverGGTestFiles,
  formatGGTestDiscoveryResults,
  formatGGTestResults,
  runGGTestFile,
  runGGTestSuite,
} from '../gg-test-runner';
export {
  afterAll,
  afterEach,
  beforeAll,
  beforeEach,
  describe,
  getActiveRegistry,
  GnosisTestRegistry,
  it,
  setActiveRegistry,
  test,
} from './registry';
export { expect } from './expect';
export {
  clearAllMocks,
  isMockFunction,
  jest,
  mock,
  restoreAllMocks,
  spyOn,
} from './mock';
export {
  readGnosisFixture,
  resolveGnosisFixturePath,
  resolveGnosisFixtureUrl,
} from './fixture';
export { buildPactSarif, runPactVerification } from './pact';
export {
  buildGnosisGgMutationReport,
  buildGnosisTypeScriptMutationReport,
  type BuildGnosisGgMutationReportInput,
  type BuildGnosisTypeScriptMutationReportInput,
  type CompilerDiagnosticLike,
  type CorrectnessViolationLike,
  type CapabilityIssueLike,
  type LeanCheckLike,
  type MutationTestingReport,
  type MutantResult,
  type MutantStatus,
} from './mutation-report';
export {
  discoverGnosisTestFiles,
  formatGnosisTestRun,
  runGnosisTests,
  runTypeScriptTestFile,
  runWrappedGgTestFile,
} from './runner';
export type {
  GnosisHookKind,
  GnosisTestCaseDefinition,
  GnosisTestEntryResult,
  GnosisTestFileResult,
  GnosisTestFn,
  GnosisTestRunOptions,
  GnosisTestRunResult,
  GnosisTestSuiteDefinition,
} from './types';
export type { ResolveGnosisFixtureOptions } from './fixture';
export {
  resolveTheoremIdsForMutant,
  type MutationTheoremMap,
} from './mutation-theorem-map';
export {
  marginalFoldWork,
  marginalFoldZeroIffHit,
  type CacheQueryResolution,
  type PositiveFoldWork,
} from './cache-observation-bridge';
export {
  UnsupportedGnosisTestFeatureError,
  unsupportedFeature,
} from './types';
