// @a0n/.ignored_gnosis/src/testing/mock.ts
import { existsSync } from 'node:fs';
import { createRequire } from 'node:module';
import path from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';
import { getActiveRegistry } from './registry';

type AnyFn = (...args: any[]) => any;
type ModuleMockFactory = () => unknown | Promise<unknown>;

export interface GnosisModuleMockContext {
  readonly defaultImporterPath: string;
  readonly runToken: string;
  readonly sourceToTranspiledUrl: ReadonlyMap<string, string>;
}

export interface GnosisMockResultRecord {
  readonly type: 'return' | 'throw';
  readonly value: unknown;
}

export interface GnosisMockState<TArgs extends unknown[] = unknown[]> {
  readonly calls: TArgs[];
  readonly results: GnosisMockResultRecord[];
}

export interface GnosisMockFunction<T extends AnyFn = AnyFn> {
  (...args: Parameters<T>): ReturnType<T>;
  readonly mock: GnosisMockState<Parameters<T>>;
  mockClear(): GnosisMockFunction<T>;
  mockReset(): GnosisMockFunction<T>;
  mockRestore(): GnosisMockFunction<T>;
  mockImplementation(fn: T): GnosisMockFunction<T>;
  mockResolvedValue(value: Awaited<ReturnType<T>>): GnosisMockFunction<T>;
  mockRejectedValue(value: unknown): GnosisMockFunction<T>;
}

export interface GnosisMockApi {
  <T extends AnyFn>(implementation?: T): GnosisMockFunction<T>;
  module(
    specifier: string,
    factory: ModuleMockFactory,
    importerPath?: string
  ): void;
}

interface GlobalMockEntry {
  clear(): void;
  restore(): void;
}

interface ActiveModuleMockContext extends GnosisModuleMockContext {
  readonly registrations: Map<string, ModuleMockFactory>;
  readonly loadedNamespaces: Map<string, Record<string, unknown>>;
}

const MOCK_REGISTRY_SYMBOL = Symbol.for('gnosis.testing.mockRegistry');
const MODULE_MOCK_CONTEXT_SYMBOL = Symbol.for('gnosis.testing.moduleMockContext');
const LOADED_MODULE_MOCKS_KEY = 'gnosis.testing.loadedModuleMocks';
const LOADED_MODULE_MOCKS_SYMBOL = Symbol.for(LOADED_MODULE_MOCKS_KEY);
const MODULE_MOCK_URL_PREFIX = 'gnode-mock://';

function getMockRegistry(): Set<GlobalMockEntry> {
  const globalState = globalThis as Record<PropertyKey, unknown>;
  const existing = globalState[MOCK_REGISTRY_SYMBOL];
  if (existing instanceof Set) {
    return existing as Set<GlobalMockEntry>;
  }
  const created = new Set<GlobalMockEntry>();
  globalState[MOCK_REGISTRY_SYMBOL] = created;
  return created;
}

function getLoadedModuleMocks(): Map<string, Record<string, unknown>> {
  const globalState = globalThis as Record<PropertyKey, unknown>;
  const existing = globalState[LOADED_MODULE_MOCKS_SYMBOL];
  if (existing instanceof Map) {
    return existing as Map<string, Record<string, unknown>>;
  }
  const created = new Map<string, Record<string, unknown>>();
  globalState[LOADED_MODULE_MOCKS_SYMBOL] = created;
  return created;
}

function getActiveModuleMockContext(): ActiveModuleMockContext | null {
  const globalState = globalThis as Record<PropertyKey, unknown>;
  const existing = globalState[MODULE_MOCK_CONTEXT_SYMBOL];
  if (
    typeof existing === 'object' &&
    existing !== null &&
    'defaultImporterPath' in existing &&
    'runToken' in existing &&
    'sourceToTranspiledUrl' in existing &&
    'registrations' in existing &&
    'loadedNamespaces' in existing
  ) {
    return existing as ActiveModuleMockContext;
  }
  return null;
}

/**
 * Handles the gnosis set Active Module Mock Context workflow.
 */
export function setActiveModuleMockContext(
  context: GnosisModuleMockContext | null
): void {
  const globalState = globalThis as Record<PropertyKey, unknown>;
  if (context === null) {
    delete globalState[MODULE_MOCK_CONTEXT_SYMBOL];
    getLoadedModuleMocks().clear();
    return;
  }

  globalState[MODULE_MOCK_CONTEXT_SYMBOL] = {
    ...context,
    registrations: new Map<string, ModuleMockFactory>(),
    loadedNamespaces: new Map<string, Record<string, unknown>>(),
  } satisfies ActiveModuleMockContext;
}

function resolveWithExtensions(basePath: string): string | null {
  const candidates = [
    basePath,
    `${basePath}.ts`,
    `${basePath}.tsx`,
    `${basePath}.mts`,
    `${basePath}.cts`,
    `${basePath}.js`,
    `${basePath}.jsx`,
    `${basePath}.mjs`,
    `${basePath}.cjs`,
    path.join(basePath, 'index.ts'),
    path.join(basePath, 'index.tsx'),
    path.join(basePath, 'index.mts'),
    path.join(basePath, 'index.cts'),
    path.join(basePath, 'index.js'),
    path.join(basePath, 'index.jsx'),
    path.join(basePath, 'index.mjs'),
    path.join(basePath, 'index.cjs'),
  ];

  for (const candidate of candidates) {
    if (existsSync(candidate)) {
      return path.resolve(candidate);
    }
  }

  return null;
}

function resolveLocalModuleSpecifier(
  specifier: string,
  importerPath: string
): string | null {
  if (specifier.startsWith('file:')) {
    return fileURLToPath(specifier);
  }

  if (!specifier.startsWith('.') && !path.isAbsolute(specifier)) {
    return null;
  }

  const basePath = path.isAbsolute(specifier)
    ? specifier
    : path.resolve(path.dirname(importerPath), specifier);

  return resolveWithExtensions(basePath);
}

function resolveRuntimeModuleUrl(
  specifier: string,
  importerPath: string
): string | null {
  const importerUrl = pathToFileURL(importerPath).href;

  try {
    const resolved = import.meta.resolve(specifier, importerUrl);
    if (typeof resolved === 'string' && resolved.length > 0) {
      return resolved;
    }
  } catch {
    // Fall through to CommonJS resolution for package imports.
  }

  try {
    const resolved = createRequire(importerPath).resolve(specifier);
    if (resolved.startsWith('node:')) {
      return resolved;
    }
    return pathToFileURL(resolved).href;
  } catch {
    return null;
  }
}

function resolveModuleMockKey(
  specifier: string,
  importerPath: string,
  context: ActiveModuleMockContext
): string {
  const localPath = resolveLocalModuleSpecifier(specifier, importerPath);
  if (localPath !== null) {
    return (
      context.sourceToTranspiledUrl.get(localPath) ?? pathToFileURL(localPath).href
    );
  }

  const runtimeUrl = resolveRuntimeModuleUrl(specifier, importerPath);
  if (runtimeUrl !== null) {
    return runtimeUrl;
  }

  throw new Error(
    `Unable to resolve mocked module '${specifier}' from '${importerPath}'.`
  );
}

function moduleMockUrl(runToken: string, resolvedUrl: string): string {
  return `${MODULE_MOCK_URL_PREFIX}${encodeURIComponent(runToken)}/${encodeURIComponent(resolvedUrl)}`;
}

function parseModuleMockUrl(
  value: string
): { readonly runToken: string; readonly resolvedUrl: string } | null {
  if (!value.startsWith(MODULE_MOCK_URL_PREFIX)) {
    return null;
  }

  const remainder = value.slice(MODULE_MOCK_URL_PREFIX.length);
  const slashIndex = remainder.indexOf('/');
  if (slashIndex === -1) {
    return null;
  }

  try {
    return {
      runToken: decodeURIComponent(remainder.slice(0, slashIndex)),
      resolvedUrl: decodeURIComponent(remainder.slice(slashIndex + 1)),
    };
  } catch {
    return null;
  }
}

function inferDefaultImporterPath(context: ActiveModuleMockContext): string {
  try {
    const activeRegistry = getActiveRegistry();
    if (activeRegistry.filePath && activeRegistry.filePath !== 'ghost.test.ts') {
      return activeRegistry.filePath;
    }
  } catch {
    // No active registry during case execution; keep the owning test file as base.
  }

  return context.defaultImporterPath;
}

function materializeMockNamespace(value: unknown): Record<string, unknown> {
  if (typeof value === 'object' && value !== null) {
    return value as Record<string, unknown>;
  }

  if (typeof value === 'function') {
    return { default: value };
  }

  return { default: value };
}

function isValidIdentifierName(value: string): boolean {
  return /^[$A-Z_a-z][$\w]*$/u.test(value);
}

/**
 * Resolves the Module Mock Redirect.
 */
export function resolveModuleMockRedirect(resolvedUrl: string): string | null {
  const context = getActiveModuleMockContext();
  if (context === null || !context.registrations.has(resolvedUrl)) {
    return null;
  }

  return moduleMockUrl(context.runToken, resolvedUrl);
}

/**
 * Loads the Module Mock Source.
 */
export function loadModuleMockSource(mockUrl: string): string | null {
  const parsed = parseModuleMockUrl(mockUrl);
  const context = getActiveModuleMockContext();
  if (
    parsed === null ||
    context === null ||
    parsed.runToken !== context.runToken
  ) {
    return null;
  }

  const factory = context.registrations.get(parsed.resolvedUrl);
  if (factory === undefined) {
    return null;
  }

  let namespace = context.loadedNamespaces.get(parsed.resolvedUrl);
  if (namespace === undefined) {
    const loaded = factory();
    if (isPromiseLike(loaded)) {
      throw new Error(
        `mock.module for '${parsed.resolvedUrl}' must return synchronously under gnode.`
      );
    }
    namespace = materializeMockNamespace(loaded);
    context.loadedNamespaces.set(parsed.resolvedUrl, namespace);
  }

  getLoadedModuleMocks().set(mockUrl, namespace);

  const namedExports = Object.keys(namespace)
    .filter((key) => key !== 'default' && isValidIdentifierName(key))
    .sort();

  const exportLines = namedExports.map(
    (name) => `export const ${name} = __module[${JSON.stringify(name)}];`
  );

  return [
    `const __registry = globalThis[Symbol.for('${LOADED_MODULE_MOCKS_KEY}')];`,
    `const __module = __registry?.get(${JSON.stringify(mockUrl)});`,
    `if (!__module) { throw new Error(${JSON.stringify(`Missing mocked module for ${parsed.resolvedUrl}`)}); }`,
    'export default __module.default ?? __module;',
    ...exportLines,
    '',
  ].join('\n');
}

function isPromiseLike(value: unknown): value is PromiseLike<unknown> {
  return (
    typeof value === 'object' &&
    value !== null &&
    'then' in value &&
    typeof (value as PromiseLike<unknown>).then === 'function'
  );
}

function createTrackedMock<T extends AnyFn>(
  implementation?: T,
  restore?: () => void
): GnosisMockFunction<T> {
  const state: {
    calls: Parameters<T>[];
    results: GnosisMockResultRecord[];
    implementation: T;
  } = {
    calls: [],
    results: [],
    implementation: implementation ?? ((() => undefined) as unknown as T),
  };

  const wrapped = function mockWrapper(
    this: unknown,
    ...args: Parameters<T>
  ): ReturnType<T> {
    state.calls.push(args);

    try {
      const result = state.implementation.apply(this, args);
      if (isPromiseLike(result)) {
        return Promise.resolve(result).then(
          (value) => {
            state.results.push({ type: 'return', value });
            return value;
          },
          (error) => {
            state.results.push({ type: 'throw', value: error });
            throw error;
          }
        ) as ReturnType<T>;
      }

      state.results.push({ type: 'return', value: result });
      return result as ReturnType<T>;
    } catch (error) {
      state.results.push({ type: 'throw', value: error });
      throw error;
    }
  } as GnosisMockFunction<T>;

  Object.defineProperty(wrapped, 'mock', {
    enumerable: true,
    value: {
      get calls() {
        return state.calls;
      },
      get results() {
        return state.results;
      },
    } satisfies GnosisMockState<Parameters<T>>,
  });

  wrapped.mockClear = () => {
    state.calls.length = 0;
    state.results.length = 0;
    return wrapped;
  };

  wrapped.mockReset = () => {
    wrapped.mockClear();
    state.implementation = (() => undefined) as unknown as T;
    return wrapped;
  };

  wrapped.mockRestore = () => {
    wrapped.mockClear();
    restore?.();
    return wrapped;
  };

  wrapped.mockImplementation = (fn: T) => {
    state.implementation = fn;
    return wrapped;
  };

  wrapped.mockResolvedValue = (value: Awaited<ReturnType<T>>) => {
    state.implementation = (() => Promise.resolve(value)) as unknown as T;
    return wrapped;
  };

  wrapped.mockRejectedValue = (value: unknown) => {
    state.implementation = (() => Promise.reject(value)) as unknown as T;
    return wrapped;
  };

  getMockRegistry().add({
    clear: () => {
      wrapped.mockClear();
    },
    restore: () => {
      wrapped.mockRestore();
    },
  });

  return wrapped;
}

function registerModuleMock(
  specifier: string,
  factory: ModuleMockFactory,
  importerPath?: string
): void {
  if (typeof specifier !== 'string' || specifier.length === 0) {
    throw new Error('mock.module specifier must be a non-empty string.');
  }
  if (typeof factory !== 'function') {
    throw new Error('mock.module factory must be a function.');
  }

  const context = getActiveModuleMockContext();
  if (context === null) {
    throw new Error(
      'mock.module requires an active Gnosis module mock context.'
    );
  }

  const resolvedImporterPath = importerPath ?? inferDefaultImporterPath(context);
  const resolvedKey = resolveModuleMockKey(
    specifier,
    resolvedImporterPath,
    context
  );

  context.registrations.set(resolvedKey, factory);
  context.loadedNamespaces.delete(resolvedKey);
}

/**
 * Returns whether is Mock Function is true.
 */
export function isMockFunction(value: unknown): value is GnosisMockFunction {
  return (
    typeof value === 'function' &&
    'mock' in value &&
    typeof (value as GnosisMockFunction).mockClear === 'function'
  );
}

export const mock = Object.assign(
  <T extends AnyFn>(implementation?: T): GnosisMockFunction<T> =>
    createTrackedMock(implementation),
  {
    module: registerModuleMock,
  }
) as GnosisMockApi;

export function spyOn<T extends object, K extends keyof T>(
  target: T,
  key: K
): GnosisMockFunction<Extract<T[K], AnyFn>> {
  const originalDescriptor = Object.getOwnPropertyDescriptor(target, key);
  const originalValue = target[key];
  if (typeof originalValue !== 'function') {
    throw new Error(
      `Cannot spy on '${String(key)}': property is not a function.`
    );
  }

  const wrapped = createTrackedMock(
    originalValue as Extract<T[K], AnyFn>,
    () => {
      if (originalDescriptor) {
        Object.defineProperty(target, key, originalDescriptor);
      } else {
        target[key] = originalValue;
      }
    }
  );

  Object.defineProperty(target, key, {
    configurable: true,
    enumerable: originalDescriptor?.enumerable ?? true,
    writable: true,
    value: wrapped,
  });

  return wrapped;
}

/**
 * Handles the gnosis clear All Mocks workflow.
 */
export function clearAllMocks(): void {
  for (const entry of getMockRegistry()) {
    entry.clear();
  }
}

/**
 * Handles the gnosis restore All Mocks workflow.
 */
export function restoreAllMocks(): void {
  for (const entry of getMockRegistry()) {
    entry.restore();
  }
}

export const jest = {
  fn: mock,
  spyOn,
  clearAllMocks,
  resetAllMocks: clearAllMocks,
  restoreAllMocks,
};
