// @a0n/.ignored_gnosis/src/testing/mutation-report.test.ts
import { describe, expect, test } from '@a0n/gnosis/test';
import {
  buildGnosisGgMutationReport,
  buildGnosisTypeScriptMutationReport,
} from './mutation-report';
import { resolveTheoremIdsForMutant } from './mutation-theorem-map';

describe('mutation-testing-elements report JSON', () => {
  test('buildGnosisGgMutationReport satisfies required schema keys', () => {
    const report = buildGnosisGgMutationReport({
      filePath: '/repo/top.gg',
      source: '(x)-[:FORK]->(y)\n',
      cwd: '/repo',
      mode: 'verify',
      compilerDiagnostics: [
        {
          line: 1,
          column: 2,
          code: 'ERR_FOO',
          message: 'example',
          severity: 'error',
        },
      ],
      correctnessViolations: [
        { kind: 'invariant', name: 'acyclic', message: 'cycle found' },
      ],
      capabilityErrors: [
        {
          severity: 'error',
          capability: 'workers',
          message: 'missing binding',
        },
      ],
      universeOk: false,
      universeReasons: ['digest mismatch'],
      buleyExceeded: true,
      buleyNumber: 99,
      maxBuley: 10,
      leanCheck: {
        attempted: true,
        ok: false,
        message: 'lean exited 1',
      },
      gnosisVersion: '1.0.0-test',
    });

    expect(report.schemaVersion).toBe('2');
    expect(report.thresholds.high).toBe(80);
    expect(report.thresholds.low).toBe(60);
    expect(report.files['top.gg']).toBeDefined();
    expect(report.files['top.gg']?.language).toBe('plaintext');
    expect(report.files['top.gg']?.source).toContain('FORK');
    expect(report.files['top.gg']?.mutants.length).toBeGreaterThan(4);
    const statuses = report.files['top.gg']?.mutants.map((m) => m.status) ?? [];
    expect(statuses).toContain('CompileError');
    expect(statuses).toContain('Killed');
  });

  test('buildGnosisTypeScriptMutationReport parses violation locations', () => {
    const report = buildGnosisTypeScriptMutationReport({
      filePath: '/repo/src/a.ts',
      source: 'export const x = 1;\n',
      cwd: '/repo',
      violations: [
        'cyclomatic /repo/src/a.ts:3 fooFn 12 > 10',
        'file-lines /repo/src/a.ts 5000 > 4000',
      ],
      gnosisVersion: '1.0.0-test',
    });

    expect(report.files['src/a.ts']).toBeDefined();
    const mutants = report.files['src/a.ts']?.mutants ?? [];
    expect(mutants).toHaveLength(2);
    expect(mutants[0]?.location.start.line).toBe(3);
    expect(mutants[0]?.mutatorName).toBe('cyclomatic');
    expect(mutants[1]?.location.start.line).toBe(1);
    expect(mutants[1]?.mutatorName).toBe('file-lines');
  });

  test('repo-relative transformation reducer keys resolve theorem ids', () => {
    const report = buildGnosisTypeScriptMutationReport({
      filePath: '/repo/src/transformation-family-formalization.ts',
      source: 'export const x = 1;\n',
      cwd: '/repo',
      violations: [
        'cyclomatic /repo/src/transformation-family-formalization.ts:12 classifyFamily 8 > 7',
      ],
      gnosisVersion: '1.0.0-test',
    });

    const fileKey = Object.keys(report.files)[0];
    expect(fileKey).toBe('src/transformation-family-formalization.ts');

    const theoremIds = resolveTheoremIdsForMutant({
      mutatorName: report.files[fileKey!]?.mutants[0]?.mutatorName ?? 'unknown',
      fileKey: fileKey ?? 'unknown',
    });

    expect(theoremIds.sort()).toEqual(
      [
        'THM-TRANSFORMATION-PROJECTION-ADDS-STRUCTURE',
        'THM-TRANSFORMATION-REINFORCEMENT-PRESERVES-STRUCTURE',
        'THM-TRANSFORMATION-DECOMPOSITION-NONINCREASING',
        'THM-TRANSFORMATION-PROJECTION-REINFORCEMENT-EXCLUSIVE',
        'THM-TRANSFORMATION-REINFORCEMENT-DECOMPOSITION-EXCLUSIVE',
        'THM-TRANSFORMATION-FEATURELESS-CANNOT-REINFORCE',
        'THM-TRANSFORMATION-DECOMPOSE-THEN-REINFORCE-REQUIRES-RESTORATION',
      ].sort()
    );
  });

  test('repo-relative decomposition runtime keys resolve theorem ids', () => {
    const report = buildGnosisTypeScriptMutationReport({
      filePath: '/repo/src/decomposition-topology.ts',
      source: 'export const x = 1;\n',
      cwd: '/repo',
      violations: [
        'cyclomatic /repo/src/decomposition-topology.ts:9 decompose 8 > 7',
      ],
      gnosisVersion: '1.0.0-test',
    });

    const fileKey = Object.keys(report.files)[0];
    expect(fileKey).toBe('src/decomposition-topology.ts');

    const theoremIds = resolveTheoremIdsForMutant({
      mutatorName: report.files[fileKey!]?.mutants[0]?.mutatorName ?? 'unknown',
      fileKey: fileKey ?? 'unknown',
    });

    expect(theoremIds).toEqual([
      'THM-TRANSFORMATION-DECOMPOSITION-NONINCREASING',
    ]);
  });
});
