// @a0n/.ignored_gnosis/src/testing/mutation-theorem-map.ts
import { dirname, isAbsolute, resolve } from 'node:path';
import { existsSync, readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';

export interface MutationTheoremMap {
  readonly version: number;
  readonly byMutatorName: Readonly<Record<string, readonly string[]>>;
  readonly byPathPrefix: Readonly<Record<string, readonly string[]>>;
}

/**
 * Resolves the Mutation Theorem Map Path.
 */
export function resolveMutationTheoremMapPath(
  moduleUrl: string | undefined = import.meta.url
): string {
  const candidates: string[] = [];

  if (typeof moduleUrl === 'string' && moduleUrl.length > 0) {
    const modulePath = moduleUrl.startsWith('file:')
      ? fileURLToPath(moduleUrl)
      : isAbsolute(moduleUrl)
        ? moduleUrl
        : resolve(process.cwd(), moduleUrl);
    const moduleDir = dirname(modulePath);
    candidates.push(resolve(moduleDir, 'data', 'mutation-theorem-map.json'));
    // Bundled layouts (e.g. dist/gnode/bridge-driver.bundle.mjs) execute from a
    // cwd outside the gnosis package; walk up from the module dir to find the
    // package root's source-tree copy so resolution is cwd-independent.
    let ancestor = moduleDir;
    for (let depth = 0; depth < 6; depth += 1) {
      candidates.push(
        resolve(ancestor, 'src/testing/data/mutation-theorem-map.json')
      );
      const parent = dirname(ancestor);
      if (parent === ancestor) {
        break;
      }
      ancestor = parent;
    }
  }

  candidates.push(
    resolve(process.cwd(), 'src/testing/data/mutation-theorem-map.json'),
    resolve(process.cwd(), '../gnosis/src/testing/data/mutation-theorem-map.json'),
    resolve(
      process.cwd(),
      'open-source/gnosis/src/testing/data/mutation-theorem-map.json'
    ),
    resolve(process.cwd(), 'dist/testing/data/mutation-theorem-map.json')
  );

  const match = candidates.find((candidate) => existsSync(candidate));
  if (match !== undefined) {
    return match;
  }

  throw new Error(
    `Unable to resolve mutation theorem map path from module url ${moduleUrl ?? '<undefined>'}`
  );
}

const DEFAULT_MAP = JSON.parse(
  readFileSync(resolveMutationTheoremMapPath(), 'utf8')
) as MutationTheoremMap;

function longestMatchingPrefix(
  fileKey: string,
  prefixes: readonly string[]
): string | undefined {
  let best: string | undefined;
  for (const p of prefixes) {
    if (fileKey === p || fileKey.startsWith(`${p}/`)) {
      if (best === undefined || p.length > best.length) {
        best = p;
      }
    }
  }
  return best;
}

/**
 * Resolves `THEOREM_LEDGER.md`-style `THM-*` ids for a mutant using the bundled map
 * (mutator name + longest matching path prefix).
 */
export function resolveTheoremIdsForMutant(
  input: {
    readonly mutatorName: string;
    readonly fileKey: string;
  },
  map: MutationTheoremMap = DEFAULT_MAP
): readonly string[] {
  const byName = map.byMutatorName[input.mutatorName];
  const prefixes = Object.keys(map.byPathPrefix);
  const matched = longestMatchingPrefix(input.fileKey, prefixes);
  const byPath = matched !== undefined ? map.byPathPrefix[matched] : undefined;
  const out = new Set<string>();
  if (byName !== undefined) {
    for (const id of byName) {
      out.add(id);
    }
  }
  if (byPath !== undefined) {
    for (const id of byPath) {
      out.add(id);
    }
  }
  return [...out];
}
