// @a0n/.ignored_gnosis/src/testing/pact.ts
import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import path from 'node:path';
import { performance } from 'node:perf_hooks';
import {
  createPactRegistry,
  type PactRegistry,
} from '../../../aeon-forge/src/deploy/pact-broker.js';
import { registerAllContracts } from '../../../aeon-forge/src/deploy/pact-contracts.js';
import {
  generatePactDocs,
  type PactScanReport,
  scanMonorepo,
} from '../../../aeon-forge/src/deploy/pact-scanner.js';
import type {
  GnosisTestEntryResult,
  GnosisTestFileResult,
  GnosisTestRunOptions,
} from './types';

interface PactFieldShape {
  readonly name: string;
  readonly type: string;
  readonly optional: boolean;
}

interface PactShape {
  readonly name: string;
  readonly producer: string;
  readonly rootHash: number;
  readonly fields: readonly PactFieldShape[];
  readonly version: number;
  readonly registeredAt: number;
}

interface PactResultShape {
  readonly name: string;
  readonly pass: boolean;
  readonly producerRoot: number;
  readonly consumerRoot: number;
  readonly divergentFields: readonly string[];
  readonly producer: string;
  readonly consumer: string;
}

interface PactVerificationShape {
  readonly allPass: boolean;
  readonly results: readonly PactResultShape[];
  readonly total: number;
  readonly passed: number;
  readonly failed: number;
  readonly verifiedAt: number;
}

type PactRegistryShape = Pick<
  PactRegistry,
  'list' | 'register' | 'save' | 'verifyAll'
>;

type PactScanReportShape = PactScanReport;

export interface GnosisPactVerificationResult {
  readonly workspaceRoot: string;
  readonly registryPath: string;
  readonly docsPath: string;
  readonly verification: PactVerificationShape;
  readonly scanReport?: PactScanReportShape;
  readonly file: GnosisTestFileResult | null;
}

const PACT_THEOREM_REFS = ['THM-VOID-BOUNDARY-SUFFICIENT-STATISTIC'] as const;
const DOCS_START_MARKER = '<!-- PACT_COVERAGE_START -->';
const DOCS_END_MARKER = '<!-- PACT_COVERAGE_END -->';

function normalizePath(value: string): string {
  return value.replace(/\\/g, '/');
}

function buildMatcher(
  pattern: string | undefined
): ((value: string) => boolean) | null {
  if (!pattern) {
    return null;
  }

  try {
    const regex = new RegExp(pattern, 'i');
    return (value: string) => regex.test(value);
  } catch {
    const lowered = pattern.toLowerCase();
    return (value: string) => value.toLowerCase().includes(lowered);
  }
}

function resolveWorkspaceRoot(startDirectory = process.cwd()): string {
  let current = path.resolve(startDirectory);

  for (;;) {
    const packageJsonPath = path.join(current, 'package.json');
    const workspaceManifestPath = path.join(current, 'pnpm-workspace.yaml');
    if (existsSync(packageJsonPath) && existsSync(workspaceManifestPath)) {
      return current;
    }

    const parent = path.dirname(current);
    if (parent === current) {
      return path.resolve(startDirectory);
    }

    current = parent;
  }
}

function updatePactDocs(docsPath: string, markdown: string): void {
  mkdirSync(path.dirname(docsPath), { recursive: true });

  if (!existsSync(docsPath)) {
    writeFileSync(docsPath, `${markdown}\n`, 'utf8');
    return;
  }

  const existing = readFileSync(docsPath, 'utf8');
  const startIndex = existing.indexOf(DOCS_START_MARKER);
  const endIndex = existing.indexOf(DOCS_END_MARKER);

  if (startIndex >= 0 && endIndex >= 0 && endIndex > startIndex) {
    const prefix = existing.slice(0, startIndex + DOCS_START_MARKER.length);
    const suffix = existing.slice(endIndex);
    writeFileSync(docsPath, `${prefix}\n\n${markdown}\n\n${suffix}`, 'utf8');
    return;
  }

  writeFileSync(docsPath, `${markdown}\n`, 'utf8');
}

function createContractEntry(result: PactResultShape): GnosisTestEntryResult {
  return {
    kind: 'contract',
    name: result.name,
    ok: result.pass,
    durationMs: 0,
    diagnostics: result.pass
      ? []
      : [
          `producer ${result.producer} <-> consumer ${result.consumer}`,
          `divergent fields: ${result.divergentFields.join(', ')}`,
        ],
    formalStatus: result.pass ? 'verified' : 'violated',
    theoremRefs: PACT_THEOREM_REFS,
    data: {
      consumer: result.consumer,
      consumerRoot: result.consumerRoot,
      divergentFields: result.divergentFields,
      producer: result.producer,
      producerRoot: result.producerRoot,
    },
  };
}

function createCoverageEntry(
  report: PactScanReportShape,
  docsUpdated: boolean,
  docsPath: string
): GnosisTestEntryResult {
  const diagnostics = [
    `coverage ${report.coveragePercent}%`,
    `covered boundaries: ${report.covered.length}`,
    `uncovered boundaries: ${report.uncovered.length}`,
    ...(report.uncovered.length > 0
      ? report.uncovered
          .slice(0, 10)
          .map(
            (entry) =>
              `${entry.typeName} from ${entry.producerPackage} (${entry.consumerCount} consumers)`
          )
      : []),
    ...(docsUpdated ? [`updated docs: ${docsPath}`] : []),
  ];

  return {
    kind: 'contract',
    name: 'coverage-scan',
    ok: true,
    durationMs: 0,
    diagnostics,
    formalStatus: 'verified',
    theoremRefs: PACT_THEOREM_REFS,
    data: {
      coveragePercent: report.coveragePercent,
      covered: report.covered.length,
      crossImports: report.crossImports.length,
      exportedTypes: report.exportedTypes.length,
      uncovered: report.uncovered.length,
    },
  };
}

function matchesPactFilter(
  entry: PactResultShape,
  matcher: ((value: string) => boolean) | null
): boolean {
  if (matcher === null) {
    return true;
  }

  return matcher(
    `${entry.name} ${entry.producer} ${
      entry.consumer
    } ${entry.divergentFields.join(' ')}`
  );
}

/**
 * Builds the Pact Sarif.
 */
export function buildPactSarif(
  verification: PactVerificationShape
): Record<string, unknown> {
  const results = verification.results
    .filter((entry) => !entry.pass)
    .map((entry) => ({
      ruleId: 'gnosis.pact.contract-drift',
      level: 'error',
      message: {
        text: `Contract "${entry.name}" failed: producer ${
          entry.producer
        } <-> consumer ${
          entry.consumer
        }. Divergent fields: ${entry.divergentFields.join(', ')}`,
      },
      locations: [
        {
          physicalLocation: {
            artifactLocation: { uri: entry.consumer },
            region: { startLine: 1 },
          },
        },
      ],
      properties: {
        consumerRoot: entry.consumerRoot,
        divergentFields: entry.divergentFields,
        producerRoot: entry.producerRoot,
      },
    }));

  return {
    version: '2.1.0',
    $schema: 'https://json.schemastore.org/sarif-2.1.0.json',
    runs: [
      {
        tool: {
          driver: {
            name: 'gnosis-pact-broker',
            version: '1.0.0',
            informationUri: 'https://mudras.us',
            rules: [
              {
                id: 'gnosis.pact.contract-drift',
                shortDescription: { text: 'Pact contract schema drift' },
              },
            ],
          },
        },
        results,
      },
    ],
  };
}

/**
 * Handles the gnosis run Pact Verification workflow.
 */
export async function runPactVerification(
  options: Pick<
    GnosisTestRunOptions,
    'grep' | 'pactScan' | 'pactUpdateDocs' | 'workspaceRoot'
  > = {}
): Promise<GnosisPactVerificationResult> {
  const started = performance.now();
  const workspaceRoot = path.resolve(
    options.workspaceRoot ?? resolveWorkspaceRoot()
  );
  const registryPath = path.join(workspaceRoot, '.aeon/pact-registry.json');
  const docsPath = path.join(
    workspaceRoot,
    'apps/docs-app/docs/contracts/pact-registry.md'
  );
  const matcher = buildMatcher(options.grep);

  // Keep verification deterministic: register contracts and consumers from the
  // source-of-truth declarations, not stale persisted consumers from earlier
  // runs. Plain verification is intentionally read-only so quality gates do not
  // churn the tracked registry metadata.
  const registry = createPactRegistry(registryPath, {
    mode: options.pactUpdateDocs ? 'update' : 'verify',
  });
  registerAllContracts(registry);
  const verification = registry.verifyAll();
  if (options.pactUpdateDocs) {
    registry.save();
  }

  const entries = verification.results
    .filter((entry) => matchesPactFilter(entry, matcher))
    .map(createContractEntry);

  let scanReport: PactScanReportShape | undefined;
  let docsUpdated = false;
  if (options.pactScan || options.pactUpdateDocs) {
    scanReport = scanMonorepo(
      workspaceRoot,
      registry.list().map((entry) => entry.name)
    );

    if (options.pactUpdateDocs) {
      updatePactDocs(docsPath, generatePactDocs(scanReport));
      docsUpdated = true;
    }

    if (
      matcher === null ||
      matcher(
        `coverage ${scanReport.coveragePercent} ${scanReport.uncovered
          .map((entry) => entry.typeName)
          .join(' ')}`
      )
    ) {
      entries.push(createCoverageEntry(scanReport, docsUpdated, docsPath));
    }
  }

  if (entries.length === 0) {
    return {
      workspaceRoot,
      registryPath,
      docsPath,
      verification,
      ...(scanReport ? { scanReport } : {}),
      file: null,
    };
  }

  const durationMs = performance.now() - started;
  const ok = entries.every((entry) => entry.ok);
  const diagnostics: string[] = [
    `contracts: ${verification.total}`,
    `passed: ${verification.passed}`,
    `failed: ${verification.failed}`,
  ];
  if (scanReport) {
    diagnostics.push(
      `coverage: ${scanReport.coveragePercent}% (${scanReport.covered.length} covered / ${scanReport.uncovered.length} uncovered)`
    );
  }

  return {
    workspaceRoot,
    registryPath,
    docsPath,
    verification,
    ...(scanReport ? { scanReport } : {}),
    file: {
      kind: 'pact',
      filePath: normalizePath(registryPath),
      ok,
      entries,
      durationMs,
      diagnostics,
      formalStatus: ok ? 'verified' : 'violated',
      theoremRefs: PACT_THEOREM_REFS,
      ...(ok
        ? {}
        : {
            formalReasons: [
              'Pact Merkle contract verification detected drift.',
            ],
          }),
    },
  };
}
