// @a0n/.ignored_gnosis/src/testing/runner.test.ts
import {
  mkdirSync,
  mkdtempSync,
  readFileSync,
  rmSync,
  writeFileSync,
} from 'node:fs';
import path from 'node:path';
import {
  discoverGnosisTestFiles,
  describe,
  expect,
  resolveGnosisFixturePath,
  runGnosisTests,
  test,
} from '@a0n/gnosis/test';
import { planGgTestLanes } from './runner';

const packageRoot = path.dirname(
  resolveGnosisFixturePath('../../package.json', import.meta.url)
);

function writeFixtureFiles(): {
  readonly tempDir: string;
  readonly tsTestPath: string;
  readonly ggTestPath: string;
} {
  const tempRoot = path.join(packageRoot, '.tmp-test-fixtures');
  mkdirSync(tempRoot, { recursive: true });
  const tempDir = mkdtempSync(path.join(tempRoot, 'gnosis-testing-runner-'));
  const tsTestPath = path.join(tempDir, 'sample.test.ts');
  const helperPath = path.join(tempDir, 'helper.ts');
  const nestedPath = path.join(tempDir, 'nested.ts');
  const ggModulePath = path.join(tempDir, 'sample.gg');
  const ggTestPath = path.join(tempDir, 'sample.test.gg');

  writeFileSync(
    nestedPath,
    [
      "import path from 'node:path';",
      '',
      'export function nestedValue(): number {',
      '  return 4;',
      '}',
      '',
      'export function nestedDirectory(): string {',
      '  return path.basename(import.meta.dir);',
      '}',
      '',
    ].join('\n'),
    'utf8'
  );

  writeFileSync(
    helperPath,
    [
      "import path from 'node:path';",
      "import { nestedDirectory, nestedValue } from './nested';",
      '',
      'export function fixtureDirectory(): string {',
      '  return `${path.basename(__dirname)}:${nestedDirectory()}`;',
      '}',
      '',
      'export function fixtureValue(): number {',
      '  return nestedValue();',
      '}',
      '',
    ].join('\n'),
    'utf8'
  );

  writeFileSync(
    tsTestPath,
    [
      "import { describe, expect, test } from '@a0n" + "/gnosis/test';",
      "import { fixtureDirectory, fixtureValue } from './helper';",
      '',
      "describe('temporary suite', () => {",
      "  test('passes through the gnosis runner', () => {",
      '    expect(fixtureValue()).toBe(4);',
      "    expect(fixtureDirectory()).toMatch('gnosis-testing-runner-');",
      '  });',
      '});',
      '',
    ].join('\n'),
    'utf8'
  );

  writeFileSync(
    ggModulePath,
    [
      "(input: Tensor { role: 'input' })",
      "(output: Tensor { role: 'output' })",
      '(input)-[:PROCESS]->(output)',
      '',
    ].join('\n'),
    'utf8'
  );

  writeFileSync(
    ggTestPath,
    [
      "(test_suite: TestSuite { name: 'temporary-topology' })",
      "(module_under_test: Verify { module: './sample.gg', beta1_max: '8' })",
      '',
      '(test_suite)-[:FORK]->(module_under_test)',
      "(module_under_test)-[:RACE { expect: 'safe' }]->(result: Result)",
      "(result)-[:FOLD { strategy: 'all-pass' }]->(verdict: Verdict)",
      '',
    ].join('\n'),
    'utf8'
  );

  return {
    tempDir,
    tsTestPath,
    ggTestPath,
  };
}


describe('Gnosis test runner', () => {
  test('runs mixed TypeScript and GG suites with one report model', async () => {
    const fixture = writeFixtureFiles();

    try {
      const result = await runGnosisTests({
        targets: [fixture.tsTestPath, fixture.ggTestPath],
        packageRoot,
        lanes: 2,
        strategy: 'cannon',
      });

      expect(result.fileCount).toBe(2);
      expect(result.files.map((file) => file.kind).sort()).toEqual([
        'gg',
        'ts',
      ]);
      expect(result.files.find((file) => file.kind === 'ts')?.ok).toBe(true);
      expect(
        result.files.find((file) => file.kind === 'gg')?.entries.length
      ).toBe(2);
    } finally {
      rmSync(fixture.tempDir, { recursive: true, force: true });
    }
  });

  test('fails closed on unsupported runner controls', async () => {
    const tempRoot = path.join(packageRoot, '.tmp-test-fixtures');
    mkdirSync(tempRoot, { recursive: true });
    const tempDir = mkdtempSync(path.join(tempRoot, 'gnosis-testing-runner-'));
    const unsupportedTestPath = path.join(tempDir, 'unsupported.test.ts');

    writeFileSync(
      unsupportedTestPath,
      [
        "import { test } from '@a0n" + "/gnosis/test';",
        '',
        "test.skip('unsupported', () => {});",
        '',
      ].join('\n'),
      'utf8'
    );

    try {
      const result = await runGnosisTests({
        targets: [unsupportedTestPath],
        packageRoot,
      });

      expect(result.ok).toBe(false);
      expect(result.files[0]?.diagnostics[0]).toMatch(
        'Unsupported Gnosis test feature'
      );
    } finally {
      rmSync(tempDir, { recursive: true, force: true });
    }
  });

  test('deduplicates overlapping targets and preserves requested report order', async () => {
    const fixture = writeFixtureFiles();

    try {
      const discovered = discoverGnosisTestFiles({
        targets: [fixture.ggTestPath, fixture.tsTestPath, fixture.tempDir],
        packageRoot,
      });
      expect(discovered.map((file) => path.basename(file.filePath))).toEqual([
        'sample.test.gg',
        'sample.test.ts',
      ]);

      const result = await runGnosisTests({
        targets: [fixture.ggTestPath, fixture.tsTestPath, fixture.tempDir],
        packageRoot,
        strategy: 'linear',
      });

      expect(result.files.map((file) => path.basename(file.filePath))).toEqual([
        'sample.test.gg',
        'sample.test.ts',
      ]);
    } finally {
      rmSync(fixture.tempDir, { recursive: true, force: true });
    }
  });

  test('plans GG lanes with duration-aware balancing instead of round robin', () => {
    const laneInputs = ['a.test.gg', 'b.test.gg', 'c.test.gg', 'd.test.gg'].map(
      (fileName) => path.join(packageRoot, '.tmp-test-fixtures', fileName)
    );

    const lanes = planGgTestLanes(laneInputs, {
      laneCount: 2,
      durationHints: {
        [laneInputs[0]!]: 9,
        [laneInputs[1]!]: 8,
        [laneInputs[2]!]: 7,
        [laneInputs[3]!]: 1,
      },
    });

    expect(lanes.lanes.map((lane) => lane.items)).toEqual([
      [laneInputs[0], laneInputs[3]],
      [laneInputs[1], laneInputs[2]],
    ]);
  });

  test('persists GG runtime history for future lane planning', async () => {
    const fixture = writeFixtureFiles();
    const historyRoot = path.join(fixture.tempDir, 'history-root');
    mkdirSync(historyRoot, { recursive: true });

    try {
      const result = await runGnosisTests({
        targets: [fixture.ggTestPath],
        packageRoot: historyRoot,
        strategy: 'linear',
      });
      expect(result.fileCount).toBe(1);

      const historyPath = path.join(
        historyRoot,
        '.tmp-test-workspaces',
        'gnode-test-history.json'
      );
      const history = JSON.parse(readFileSync(historyPath, 'utf8')) as {
        readonly files: Record<
          string,
          {
            readonly emaDurationMs: number;
            readonly lastDurationMs: number;
            readonly runs: number;
          }
        >;
      };
      const entry = history.files[fixture.ggTestPath];

      expect(entry.emaDurationMs).toBeGreaterThan(0);
      expect(entry.lastDurationMs).toBeGreaterThan(0);
      expect(entry.runs).toBe(1);
    } finally {
      rmSync(fixture.tempDir, { recursive: true, force: true });
    }
  });
});
