// @a0n/.ignored_gnosis/src/testing/runner.ts
import {
  existsSync,
  mkdirSync,
  readdirSync,
  readFileSync,
  realpathSync,
  rmSync,
  statSync,
  writeFileSync,
} from 'node:fs';
import path from 'node:path';
import { tmpdir } from 'node:os';
import { builtinModules, createRequire } from 'node:module';
import { performance } from 'node:perf_hooks';
import { fileURLToPath, pathToFileURL } from 'node:url';
import { createHash } from 'node:crypto';
import { transform } from '@a0n/a0/bundler';
import * as ts from 'typescript';
import { BettyCompiler } from '../betty/compiler';
import { analyzeGnosisSource } from '../analysis';
import { runGGTestFile } from '../gg-test-runner';
import { GnosisEngine } from '../runtime/engine';
import { GnosisRegistry } from '../runtime/registry';
import {
  afterAll,
  afterEach,
  beforeAll,
  beforeEach,
  describe,
  GnosisTestRegistry,
  it,
  setActiveRegistry,
  test,
} from './registry';
import { expect } from './expect';
import {
  readGnosisFixture,
  resolveGnosisFixturePath,
  resolveGnosisFixtureUrl,
} from './fixture';
import {
  clearAllMocks,
  jest,
  loadModuleMockSource,
  mock,
  resolveModuleMockRedirect,
  restoreAllMocks,
  setActiveModuleMockContext,
  spyOn,
} from './mock';
import { buildPactSarif, runPactVerification } from './pact';
import type {
  GnosisTestCaseDefinition,
  GnosisTestEntryResult,
  GnosisTestFileResult,
  GnosisTestFn,
  GnosisTestRunOptions,
  GnosisTestRunResult,
  GnosisTestSuiteDefinition,
} from './types';
import { unsupportedFeature } from './types';

type Strategy = 'cannon' | 'linear';
type FormalStatus = 'verified' | 'open' | 'violated';

interface ModuleResolveHookContext {
  readonly conditions?: readonly string[];
  readonly importAttributes?: Record<string, string>;
  readonly parentURL?: string;
}

interface ModuleResolveResult {
  readonly format?: string;
  readonly importAttributes?: Record<string, string>;
  readonly shortCircuit?: boolean;
  readonly url: string;
}

interface ModuleLoadHookContext {
  readonly format?: string;
  readonly importAttributes?: Record<string, string>;
}

interface ModuleLoadResult {
  readonly format: string;
  readonly shortCircuit?: boolean;
  readonly source?: string | ArrayBuffer | ArrayBufferView;
}

interface ModuleHooksHandle {
  deregister(): void;
}

type RegisterHooksFn = (hooks: {
  readonly resolve?: (
    specifier: string,
    context: ModuleResolveHookContext,
    nextResolve: (
      specifier: string,
      context: ModuleResolveHookContext
    ) => ModuleResolveResult
  ) => ModuleResolveResult;
  readonly load?: (
    url: string,
    context: ModuleLoadHookContext,
    nextLoad: (url: string, context: ModuleLoadHookContext) => ModuleLoadResult
  ) => ModuleLoadResult;
}) => ModuleHooksHandle;

function resolveRegisterHooksRequireBase(): string {
  if (typeof __filename === 'string' && __filename.length > 0) {
    return __filename;
  }

  const moduleUrl = (import.meta as { url?: string }).url;
  if (typeof moduleUrl === 'string' && moduleUrl.length > 0) {
    return moduleUrl;
  }

  return path.resolve(process.cwd(), 'package.json');
}

const { registerHooks } = createRequire(resolveRegisterHooksRequireBase())(
  'node:module'
) as {
  readonly registerHooks: RegisterHooksFn;
};

interface DiscoveredTestFile {
  readonly kind: 'ts' | 'gg';
  readonly filePath: string;
  readonly index: number;
}

interface RotationResult<T> {
  readonly lanes: readonly {
    readonly laneIndex: number;
    readonly items: readonly T[];
    readonly estimatedWeight?: number;
  }[];
}

interface ExecutionStep {
  readonly kind: 'beforeAll' | 'afterAll' | 'beforeEach' | 'afterEach' | 'case';
  readonly name: string;
  readonly fn: GnosisTestFn;
}

interface ExecutionPayload {
  failure: {
    readonly stepName: string;
    readonly error: unknown;
  } | null;
}

interface SuiteRunState {
  readonly filePath: string;
  readonly runAllCases: boolean;
  readonly matcher: ((value: string) => boolean) | null;
  readonly bail: boolean;
  shouldStop: boolean;
}

const TS_TEST_PATTERN = /\.(test|spec)\.(ts|tsx)$/;
const GG_TEST_PATTERN = /\.test\.gg$/;
const EXCLUDED_DIRS = new Set([
  '.git',
  'coverage',
  'dist',
  'node_modules',
  '.tmp-test-workspaces',
]);
const GG_RUNTIME_HISTORY_PATH = path.join(
  '.tmp-test-workspaces',
  'gnode-test-history.json'
);
const TEST_RUNTIME_KEY = 'gnosis.testing.runtimeExports';
const TEST_RUNTIME_SYMBOL = Symbol.for(TEST_RUNTIME_KEY);

interface GgRuntimeHistoryEntry {
  readonly emaDurationMs: number;
  readonly lastDurationMs: number;
  readonly runs: number;
  readonly updatedAt: string;
}

interface GgRuntimeHistoryFile {
  readonly version: 1;
  readonly files: Record<string, GgRuntimeHistoryEntry>;
}

function combineFormalStatus(
  statuses: readonly (FormalStatus | undefined)[]
): FormalStatus {
  if (statuses.some((status) => status === 'violated')) {
    return 'violated';
  }
  if (statuses.some((status) => status === 'open')) {
    return 'open';
  }
  return 'verified';
}

function collectTheoremRefs(
  values: readonly (readonly string[] | undefined)[]
): readonly string[] {
  return [...new Set(values.flatMap((value) => value ?? []))].sort();
}

function defaultPackageRoot(): string {
  return path.resolve(path.dirname(fileURLToPath(import.meta.url)), '../..');
}

function normalizePath(value: string): string {
  return value.replace(/\\/g, '/');
}

function isTestSourceFile(filePath: string): boolean {
  return TS_TEST_PATTERN.test(filePath);
}

function isGgSuiteFile(filePath: string): boolean {
  return GG_TEST_PATTERN.test(filePath);
}

function sortPathsLexicographically(values: readonly string[]): string[] {
  return [...values].sort((left, right) => left.localeCompare(right));
}

function buildMatcher(
  pattern: string | undefined
): ((value: string) => boolean) | null {
  if (!pattern) {
    return null;
  }

  try {
    const regex = new RegExp(pattern, 'i');
    return (value: string) => regex.test(value);
  } catch {
    const normalized = pattern.toLowerCase();
    return (value: string) => value.toLowerCase().includes(normalized);
  }
}

function discoverFromTarget(
  target: string,
  results: DiscoveredTestFile[],
  nextIndex: () => number,
  seen: Set<string>
): void {
  const resolved = path.resolve(target);
  if (!existsSync(resolved)) {
    return;
  }

  const stat = statSync(resolved);
  if (stat.isFile()) {
    if (isTestSourceFile(resolved)) {
      if (!seen.has(resolved)) {
        seen.add(resolved);
        results.push({ kind: 'ts', filePath: resolved, index: nextIndex() });
      }
    } else if (isGgSuiteFile(resolved)) {
      if (!seen.has(resolved)) {
        seen.add(resolved);
        results.push({ kind: 'gg', filePath: resolved, index: nextIndex() });
      }
    }
    return;
  }

  const entries = readdirSync(resolved, { withFileTypes: true }).sort(
    (left, right) => left.name.localeCompare(right.name)
  );
  for (const entry of entries) {
    if (entry.isDirectory()) {
      if (EXCLUDED_DIRS.has(entry.name)) {
        continue;
      }
      discoverFromTarget(
        path.join(resolved, entry.name),
        results,
        nextIndex,
        seen
      );
      continue;
    }

    const entryPath = path.join(resolved, entry.name);
    if (isTestSourceFile(entryPath)) {
      if (!seen.has(entryPath)) {
        seen.add(entryPath);
        results.push({ kind: 'ts', filePath: entryPath, index: nextIndex() });
      }
    } else if (isGgSuiteFile(entryPath)) {
      if (!seen.has(entryPath)) {
        seen.add(entryPath);
        results.push({ kind: 'gg', filePath: entryPath, index: nextIndex() });
      }
    }
  }
}

/**
 * Handles the gnosis discover Gnosis Test Files workflow.
 */
export function discoverGnosisTestFiles(
  options: Pick<GnosisTestRunOptions, 'packageRoot' | 'roots' | 'targets'> = {}
): DiscoveredTestFile[] {
  const packageRoot = path.resolve(options.packageRoot ?? defaultPackageRoot());
  const roots =
    options.targets !== undefined
      ? options.targets.map((target) =>
          path.isAbsolute(target) ? target : path.resolve(packageRoot, target)
        )
      : options.roots?.length && options.roots.length > 0
      ? options.roots.map((root) => path.resolve(packageRoot, root))
      : [
          path.resolve(packageRoot, 'src'),
          path.resolve(packageRoot, 'topologies'),
        ];
  const results: DiscoveredTestFile[] = [];
  const seen = new Set<string>();
  let cursor = 0;
  const nextIndex = () => {
    const current = cursor;
    cursor += 1;
    return current;
  };

  for (const root of roots) {
    discoverFromTarget(root, results, nextIndex, seen);
  }

  return [...results].sort((left, right) => left.index - right.index);
}

function ggRuntimeHistoryPath(packageRoot: string): string {
  return path.join(packageRoot, GG_RUNTIME_HISTORY_PATH);
}

function loadGgRuntimeHistory(
  packageRoot: string
): Map<string, GgRuntimeHistoryEntry> {
  const historyPath = ggRuntimeHistoryPath(packageRoot);
  if (!existsSync(historyPath)) {
    return new Map();
  }

  try {
    const parsed = JSON.parse(
      readFileSync(historyPath, 'utf8')
    ) as Partial<GgRuntimeHistoryFile>;
    if (
      parsed.version !== 1 ||
      typeof parsed.files !== 'object' ||
      parsed.files === null
    ) {
      return new Map();
    }

    return new Map(
      Object.entries(parsed.files).flatMap(([filePath, value]) => {
        if (
          typeof value?.emaDurationMs !== 'number' ||
          typeof value?.lastDurationMs !== 'number' ||
          typeof value?.runs !== 'number' ||
          typeof value?.updatedAt !== 'string'
        ) {
          return [];
        }

        return [[filePath, value]];
      })
    );
  } catch {
    return new Map();
  }
}

function saveGgRuntimeHistory(
  packageRoot: string,
  history: ReadonlyMap<string, GgRuntimeHistoryEntry>
): void {
  const historyPath = ggRuntimeHistoryPath(packageRoot);
  mkdirSync(path.dirname(historyPath), { recursive: true });
  writeFileSync(
    historyPath,
    `${JSON.stringify(
      {
        version: 1,
        files: Object.fromEntries(
          [...history.entries()].sort(([left], [right]) =>
            left.localeCompare(right)
          )
        ),
      } satisfies GgRuntimeHistoryFile,
      null,
      2
    )}\n`,
    'utf8'
  );
}

function updateGgRuntimeHistoryEntry(
  previous: GgRuntimeHistoryEntry | undefined,
  durationMs: number
): GgRuntimeHistoryEntry {
  const roundedDurationMs = Number(durationMs.toFixed(3));
  if (previous === undefined) {
    return {
      emaDurationMs: roundedDurationMs,
      lastDurationMs: roundedDurationMs,
      runs: 1,
      updatedAt: new Date().toISOString(),
    };
  }

  return {
    emaDurationMs: Number(
      (previous.emaDurationMs * 0.7 + roundedDurationMs * 0.3).toFixed(3)
    ),
    lastDurationMs: roundedDurationMs,
    runs: previous.runs + 1,
    updatedAt: new Date().toISOString(),
  };
}

function resolveRelativeTopologyModule(
  importerPath: string,
  specifier: string
): string | null {
  if (!isRelativeModuleSpecifier(specifier)) {
    return null;
  }

  const resolved = path.resolve(path.dirname(importerPath), specifier);
  if (!existsSync(resolved)) {
    return null;
  }

  return statSync(resolved).isFile() ? resolved : null;
}

function statTextFile(filePath: string): {
  readonly bytes: number;
  readonly lines: number;
} {
  const text = readFileSync(filePath, 'utf8');
  return {
    bytes: Buffer.byteLength(text),
    lines: text.split('\n').length,
  };
}

function estimateGgStaticWeight(filePath: string): number {
  const source = readFileSync(filePath, 'utf8');
  const localStats = {
    bytes: Buffer.byteLength(source),
    lines: source.split('\n').length,
  };
  const moduleSpecifiers = [
    ...source.matchAll(/\bmodule\s*:\s*'([^']+)'/g),
  ].map((match) => match[1]!);
  const modulePaths = sortPathsLexicographically(
    moduleSpecifiers
      .map((specifier) => resolveRelativeTopologyModule(filePath, specifier))
      .filter((value): value is string => value !== null)
  );
  const moduleStats = modulePaths.reduce(
    (aggregate, modulePath) => {
      const moduleFile = statTextFile(modulePath);
      return {
        bytes: aggregate.bytes + moduleFile.bytes,
        lines: aggregate.lines + moduleFile.lines,
      };
    },
    { bytes: 0, lines: 0 }
  );
  const verifyCount = (source.match(/\bVerify\b/g) ?? []).length;
  const raceCount = (source.match(/\bRACE\b/g) ?? []).length;
  const foldCount = (source.match(/\bFOLD\b/g) ?? []).length;

  return (
    localStats.bytes +
    moduleStats.bytes +
    (localStats.lines + moduleStats.lines) * 8 +
    modulePaths.length * 512 +
    verifyCount * 1024 +
    raceCount * 256 +
    foldCount * 128
  );
}

/**
 * Handles the gnosis plan Gg Test Lanes workflow.
 */
export function planGgTestLanes(
  filePaths: readonly string[],
  options: {
    readonly laneCount: number;
    readonly packageRoot?: string;
    readonly durationHints?: Readonly<Record<string, number>>;
  }
): RotationResult<string> {
  const normalizedLaneCount = Math.max(1, options.laneCount);
  const history =
    options.durationHints !== undefined
      ? new Map(Object.entries(options.durationHints))
      : options.packageRoot
      ? new Map(
          [...loadGgRuntimeHistory(options.packageRoot).entries()].map(
            ([filePath, entry]) => [filePath, entry.emaDurationMs]
          )
        )
      : new Map<string, number>();
  const candidates = filePaths.map((filePath, index) => ({
    filePath,
    index,
    weight: history.get(filePath) ?? estimateGgStaticWeight(filePath),
  }));
  const lanes = Array.from({ length: normalizedLaneCount }, (_, laneIndex) => ({
    laneIndex,
    items: [] as string[],
    estimatedWeight: 0,
  }));

  for (const candidate of candidates.sort((left, right) => {
    if (right.weight !== left.weight) {
      return right.weight - left.weight;
    }
    return left.index - right.index;
  })) {
    lanes.sort((left, right) => {
      if (left.estimatedWeight !== right.estimatedWeight) {
        return left.estimatedWeight - right.estimatedWeight;
      }
      return left.laneIndex - right.laneIndex;
    });

    lanes[0]!.items.push(candidate.filePath);
    lanes[0]!.estimatedWeight += candidate.weight;
  }

  return {
    lanes: [...lanes]
      .sort((left, right) => left.laneIndex - right.laneIndex)
      .map((lane) => ({
        laneIndex: lane.laneIndex,
        items: lane.items,
        estimatedWeight: lane.estimatedWeight,
      })),
  };
}

function loaderForPath(filePath: string): 'ts' | 'tsx' | 'js' | 'jsx' {
  if (filePath.endsWith('.tsx')) {
    return 'tsx';
  }
  if (filePath.endsWith('.jsx')) {
    return 'jsx';
  }
  if (filePath.endsWith('.mts') || filePath.endsWith('.cts')) {
    return 'ts';
  }
  if (
    filePath.endsWith('.js') ||
    filePath.endsWith('.mjs') ||
    filePath.endsWith('.cjs')
  ) {
    return 'js';
  }
  return 'ts';
}

function canonicalFileUrl(filePath: string): string {
  try {
    return pathToFileURL(realpathSync.native(filePath)).href;
  } catch {
    return pathToFileURL(filePath).href;
  }
}

// Globals replaced via vi.stubGlobal, so vi.unstubAllGlobals can restore them.
// Without this the stub silently no-ops and tests that mock fetch/globals leak
// through to the real implementation (e.g. live network).
const stubbedGlobals = new Map<string, { existed: boolean; value: unknown }>();

function createVitestRuntime() {
  const runtime = Object.freeze({
    fn: mock,
    spyOn,
    clearAllMocks,
    restoreAllMocks,
    stubGlobal(name: string, value: unknown) {
      const g = globalThis as Record<string, unknown>;
      if (!stubbedGlobals.has(name)) {
        stubbedGlobals.set(name, { existed: name in g, value: g[name] });
      }
      g[name] = value;
      return runtime;
    },
    unstubAllGlobals() {
      const g = globalThis as Record<string, unknown>;
      for (const [name, prev] of stubbedGlobals) {
        if (prev.existed) g[name] = prev.value;
        else delete g[name];
      }
      stubbedGlobals.clear();
      return runtime;
    },
    mocked<T>(value: T): T {
      return value;
    },
    useFakeTimers(): never {
      return unsupportedFeature('vi.useFakeTimers');
    },
    useRealTimers(): never {
      return unsupportedFeature('vi.useRealTimers');
    },
    advanceTimersByTime(): never {
      return unsupportedFeature('vi.advanceTimersByTime');
    },
    runAllTimers(): never {
      return unsupportedFeature('vi.runAllTimers');
    },
    mock(
      specifier: string,
      factory: () => unknown | Promise<unknown>,
      importerPath?: string
    ): void {
      // vitest's vi.mock — delegates to the canonical module-mock registry.
      return mock.module(specifier, factory, importerPath);
    },
    module(
      specifier: string,
      factory: () => unknown | Promise<unknown>,
      importerPath?: string
    ): void {
      return mock.module(specifier, factory, importerPath);
    },
  });
  return runtime;
}

function setTestRuntimeExports(): void {
  (globalThis as Record<PropertyKey, unknown>)[TEST_RUNTIME_SYMBOL] = {
    afterAll,
    afterEach,
    buildPactSarif,
    beforeAll,
    beforeEach,
    clearAllMocks,
    discoverGnosisTestFiles,
    describe,
    expect,
    formatGnosisTestRun,
    it,
    jest,
    mock,
    readGnosisFixture,
    runGnosisTests,
    runPactVerification,
    runTypeScriptTestFile,
    runWrappedGgTestFile,
    resolveGnosisFixturePath,
    resolveGnosisFixtureUrl,
    restoreAllMocks,
    spyOn,
    test,
    vi: createVitestRuntime(),
  };
}

function clearTestRuntimeExports(): void {
  delete (globalThis as Record<PropertyKey, unknown>)[TEST_RUNTIME_SYMBOL];
}

function runtimeShimSource(
  namedExports: readonly string[],
  includeVi = false
): string {
  const exportLines = namedExports
    .map((name) => `export const ${name} = runtime.${name};`)
    .join('\n');

  return [
    `const runtime = globalThis[Symbol.for('${TEST_RUNTIME_KEY}')];`,
    "if (!runtime) { throw new Error('No active Gnosis test runtime exports.'); }",
    exportLines,
    ...(includeVi ? ['export const vi = runtime.vi;'] : []),
    '',
  ].join('\n');
}

/**
 * Source for a generated `.mjs` shim that stands in for a `.wasm` asset import.
 *
 * App/runtime code imports a `.wasm` asset (often via a `?module` suffix)
 * expecting the bundler to hand back a precompiled `WebAssembly.Module`
 * (Cloudflare Workers semantics). Node's experimental Wasm-ESM integration
 * instead tries to ESM-instantiate the `.wasm` and resolve its declared host
 * imports (`(import "env" "expf" ...)`) as packages, which fails with
 * ERR_MODULE_NOT_FOUND ('env'). The `.wasm` is meant to be loaded as BYTES and
 * instantiated with an explicit imports object by the runtime, never via
 * Wasm-ESM.
 *
 * This shim reads the asset bytes from disk and synchronously compiles them to
 * a `WebAssembly.Module`, exporting it as the default export — matching the
 * bundler contract (`import mod from './k.wasm?module'` ⇒ `WebAssembly.Module`).
 * It also exports the raw bytes for callers that load via `readFileSync`/fetch.
 * No host imports are resolved at module-load time, so the failing Wasm-ESM
 * path is never taken.
 */
function wasmAssetShimSource(wasmFilePath: string): string {
  const wasmHref = canonicalFileUrl(wasmFilePath);
  return [
    "import { readFileSync } from 'node:fs';",
    "import { fileURLToPath } from 'node:url';",
    `const wasmUrl = ${JSON.stringify(wasmHref)};`,
    'const wasmBytes = readFileSync(fileURLToPath(wasmUrl));',
    'const wasmModule = new WebAssembly.Module(wasmBytes);',
    'export const bytes = wasmBytes;',
    'export const url = wasmUrl;',
    'export const module = wasmModule;',
    'export default wasmModule;',
    '',
  ].join('\n');
}

/**
 * Emit (once) a generated `.mjs` shim for a `.wasm` asset and return its path.
 * The shim is shared process-wide (keyed by the asset path) so the Node ESM
 * loader caches a single instance across every test file in the run.
 */
function ensureWasmAssetShim(
  wasmFilePath: string,
  createdPaths: Set<string>
): string {
  const cached = SHARED_SOURCE_MODULE_CACHE.get(wasmFilePath);
  if (cached !== undefined) {
    createdPaths.add(cached);
    return cached;
  }

  const roots = createSharedArtifactRoots();
  const shimPath = path.join(
    roots.moduleRoot,
    `.__gnode_wasm_shim__${createHash('sha1')
      .update(wasmFilePath)
      .digest('hex')
      .slice(0, 16)}.mjs`
  );
  writeFileSync(shimPath, wasmAssetShimSource(wasmFilePath), 'utf8');
  SHARED_SOURCE_MODULE_CACHE.set(wasmFilePath, shimPath);
  SHARED_SOURCE_MODULE_PATHS.add(shimPath);
  createdPaths.add(shimPath);
  return shimPath;
}

function createRunToken(filePath: string): string {
  return createHash('sha1')
    .update(`${filePath}:${process.pid}:${Date.now()}:${Math.random()}`)
    .digest('hex')
    .slice(0, 16);
}

const packageRootByPathCache = new Map<string, string | null>();

/**
 * Process-wide cache of transpiled source dependency module paths,
 * keyed by absolute source file path.
 *
 * Each `gnode test` invocation visits hundreds of test files. Every test
 * file pulls in dozens of shared source modules. Without process-wide
 * deduplication, every test file produced a fresh transpiled `.mjs` for
 * each transitive dependency (because the modulePath hash was salted with
 * the per-file `runToken`), and the Node ESM loader retained every one of
 * them. The resulting heap pressure produced the OOMs observed at
 * package-level test runs. Sharing transpiled dependency artifacts is
 * safe: the runtime shims (`@a0n/gnosis/test`, vitest, bun) resolve the
 * active registry through `globalThis[Symbol.for(...)]` at runtime, so
 * the same transpiled dep can serve every test file in the run.
 */
const SHARED_SOURCE_MODULE_CACHE = new Map<string, string>();
const SHARED_SOURCE_MODULE_PATHS = new Set<string>();
let sharedShimCache: {
  readonly bunShimPath: string;
  readonly indexShimPath: string;
  readonly vitestShimPath: string;
} | null = null;

function findNearestPackageRoot(filePath: string): string | null {
  const cached = packageRootByPathCache.get(filePath);
  if (cached !== undefined) {
    return cached;
  }

  let cursor = path.dirname(filePath);
  while (true) {
    const candidate = path.join(cursor, 'package.json');
    if (existsSync(candidate)) {
      packageRootByPathCache.set(filePath, cursor);
      return cursor;
    }

    const parent = path.dirname(cursor);
    if (parent === cursor) {
      packageRootByPathCache.set(filePath, null);
      return null;
    }
    cursor = parent;
  }
}

/**
 * Transpiled `.mjs` and shims for one `gnode test` load. Lives under **system temp** (not the repo)
 * so `a0` low-disk `pruneWorkspaceTempDirs` cannot delete live imports (ENOENT on `.__gnode_test_dep__*.mjs`).
 *
 * @param _artifactBaseRoot — legacy; kept for call-site compatibility (was nearest package root).
 */
function createTransientArtifactRoots(
  _artifactBaseRoot: string,
  runToken: string
): {
  readonly artifactRoot: string;
  readonly moduleRoot: string;
  readonly shimRoot: string;
} {
  const artifactRoot = path.join(
    tmpdir(),
    'gnode-test-runtime',
    `.__gnode_test_runtime__${runToken}`
  );
  const moduleRoot = path.join(artifactRoot, 'modules');
  const shimRoot = path.join(artifactRoot, 'shims');
  mkdirSync(moduleRoot, { recursive: true });
  mkdirSync(shimRoot, { recursive: true });
  return {
    artifactRoot,
    moduleRoot,
    shimRoot,
  };
}

function ensureRuntimeShims(
  _shimRoot: string
): {
  readonly bunShimPath: string;
  readonly indexShimPath: string;
  readonly vitestShimPath: string;
} {
  // Shims are pure global-lookup wrappers around the active test runtime,
  // so they can be reused for the lifetime of the process. Sharing them
  // is critical: the transpiled source dependencies embed shim paths in
  // their `import` statements, and the dependency cache is process-wide.
  if (sharedShimCache !== null) {
    return sharedShimCache;
  }

  // Always emit shims into the shared shim root so paths embedded in
  // shared transpiled dependencies remain valid across the whole run.
  const shimRoot = createSharedArtifactRoots().shimRoot;
  mkdirSync(shimRoot, { recursive: true });

  const namedExports = [
    'afterAll',
    'afterEach',
    'buildPactSarif',
    'beforeAll',
    'beforeEach',
    'describe',
    'discoverGnosisTestFiles',
    'formatGnosisTestRun',
    'expect',
    'it',
    'jest',
    'mock',
    'readGnosisFixture',
    'runGnosisTests',
    'runPactVerification',
    'runTypeScriptTestFile',
    'runWrappedGgTestFile',
    'resolveGnosisFixturePath',
    'resolveGnosisFixtureUrl',
    'restoreAllMocks',
    'spyOn',
    'test',
  ] as const;
  const indexShimPath = path.join(shimRoot, 'index.mjs');
  const bunShimPath = path.join(shimRoot, 'bun.mjs');
  const vitestShimPath = path.join(shimRoot, 'vitest.mjs');

  writeFileSync(indexShimPath, runtimeShimSource(namedExports), 'utf8');
  writeFileSync(bunShimPath, runtimeShimSource(namedExports), 'utf8');
  writeFileSync(vitestShimPath, runtimeShimSource(namedExports, true), 'utf8');

  sharedShimCache = {
    bunShimPath,
    indexShimPath,
    vitestShimPath,
  };
  return sharedShimCache;
}

function runtimeCompatSpecifier(
  specifier: string,
  shims: {
    readonly bunShimPath: string;
    readonly indexShimPath: string;
    readonly vitestShimPath: string;
  }
): string | null {
  const bunShimHref = pathToFileURL(shims.bunShimPath).href;
  const indexShimHref = pathToFileURL(shims.indexShimPath).href;
  const vitestShimHref = pathToFileURL(shims.vitestShimPath).href;

  if (
    specifier === 'bun:test' ||
    specifier === '@a0n/gnosis/test/bun' ||
    specifier === '@a0n/gnosis/test/bun.js'
  ) {
    return bunShimHref;
  }
  if (
    specifier === 'vitest' ||
    specifier === '@a0n/gnosis/test/vitest' ||
    specifier === '@a0n/gnosis/test/vitest.js'
  ) {
    return vitestShimHref;
  }
  if (
    specifier === '@jest/globals' ||
    specifier === '@a0n/gnosis/test' ||
    specifier === '@a0n/gnosis/test.js' ||
    specifier === '@a0n/gnosis/test/index.js'
  ) {
    return indexShimHref;
  }

  return null;
}

function isRelativeModuleSpecifier(specifier: string): boolean {
  return specifier.startsWith('./') || specifier.startsWith('../');
}

function isTranspilableSourceFile(filePath: string): boolean {
  if (/\.d\.[cm]?ts$/.test(filePath)) {
    return false;
  }
  return /\.(?:[cm]?ts|tsx|[cm]?js|jsx)$/.test(filePath);
}

function isWasmAssetFile(filePath: string): boolean {
  return filePath.endsWith('.wasm');
}

/**
 * Strip a trailing `?query` / `#hash` suffix from a module specifier.
 *
 * Bundlers (Cloudflare/esbuild/Vite) attach query suffixes such as
 * `./kernels.wasm?module` to select an asset-loading mode. The suffix is not
 * part of the on-disk path, so it must be removed before filesystem
 * resolution. The original specifier (with suffix) is still useful elsewhere
 * to detect the requested asset mode, so this only strips for resolution.
 */
function stripModuleSpecifierSuffix(specifier: string): string {
  const queryIndex = specifier.search(/[?#]/);
  return queryIndex === -1 ? specifier : specifier.slice(0, queryIndex);
}

function firstExistingSourceFile(candidates: readonly string[]): string | null {
  for (const candidate of candidates) {
    if (!existsSync(candidate)) {
      continue;
    }

    const stat = statSync(candidate);
    if (stat.isFile() && isTranspilableSourceFile(candidate)) {
      return candidate;
    }
  }

  return null;
}

const RUNTIME_SOURCE_EXTENSIONS = new Set([
  '.ts', '.tsx', '.mts', '.cts', '.js', '.jsx', '.mjs', '.cjs',
]);

function resolveWithExtensions(resolved: string): string | null {
  const extension = path.extname(resolved);
  if (extension.length > 0 && existsSync(resolved)) {
    const stat = statSync(resolved);
    if (stat.isFile()) {
      if (isWasmAssetFile(resolved) || isTranspilableSourceFile(resolved)) {
        return resolved;
      }
    }
  }

  // Append source extensions when the specifier is extensionless, OR when its
  // apparent "extension" is not a real source extension and no such file exists
  // — e.g. `./storyWitnesses.generated` (extname `.generated`) is really a base
  // name needing `.ts` appended. Without this, dotted-name modules never resolve.
  if (
    extension.length === 0 ||
    (!RUNTIME_SOURCE_EXTENSIONS.has(extension) && !existsSync(resolved))
  ) {
    const appended = firstExistingSourceFile([
      `${resolved}.ts`,
      `${resolved}.tsx`,
      `${resolved}.mts`,
      `${resolved}.cts`,
      `${resolved}.js`,
      `${resolved}.jsx`,
      `${resolved}.mjs`,
      `${resolved}.cjs`,
    ]);
    if (appended) return appended;
    if (extension.length === 0) {
      return firstExistingSourceFile([
        path.join(resolved, 'index.ts'),
        path.join(resolved, 'index.tsx'),
        path.join(resolved, 'index.mts'),
        path.join(resolved, 'index.cts'),
        path.join(resolved, 'index.js'),
        path.join(resolved, 'index.jsx'),
        path.join(resolved, 'index.mjs'),
        path.join(resolved, 'index.cjs'),
      ]);
    }
  }

  const stem = resolved.slice(0, resolved.length - extension.length);
  return firstExistingSourceFile(
    extension === '.mjs'
      ? [`${stem}.mts`, `${stem}.ts`, `${stem}.tsx`]
      : extension === '.cjs'
      ? [`${stem}.cts`, `${stem}.ts`, `${stem}.tsx`]
      : extension === '.jsx'
      ? [`${stem}.tsx`, `${stem}.ts`]
      : extension === '.js'
      ? [`${stem}.ts`, `${stem}.tsx`, `${stem}.mts`, `${stem}.cts`]
      : []
  );
}

const MONOREPO_BOUNDARIES = [
  '/open-source/',
  '/apps/',
  '/shared-ui/',
  '/shared-utils/',
  '/packages/',
] as const;

const tsconfigPathsCache = new Map<string, readonly [string, readonly string[]][]>();
const NODE_BUILTIN_MODULES = new Set([
  ...builtinModules,
  ...builtinModules.map((name) => `node:${name}`),
]);

function findMonorepoRoot(importerPath: string): string | null {
  for (const boundary of MONOREPO_BOUNDARIES) {
    const index = importerPath.indexOf(boundary);
    if (index !== -1) {
      return importerPath.slice(0, index);
    }
  }
  return null;
}

function matchTsconfigPathPattern(
  pattern: string,
  specifier: string
): string | null {
  const wildcardIndex = pattern.indexOf('*');
  if (wildcardIndex === -1) {
    return pattern === specifier ? '' : null;
  }

  const prefix = pattern.slice(0, wildcardIndex);
  const suffix = pattern.slice(wildcardIndex + 1);
  if (!specifier.startsWith(prefix) || !specifier.endsWith(suffix)) {
    return null;
  }

  return specifier.slice(prefix.length, specifier.length - suffix.length);
}

function applyTsconfigTargetPattern(target: string, wildcard: string): string {
  const wildcardIndex = target.indexOf('*');
  if (wildcardIndex === -1) {
    return target;
  }
  return `${target.slice(0, wildcardIndex)}${wildcard}${target.slice(
    wildcardIndex + 1
  )}`;
}

function readTsconfigPathEntries(
  monorepoRoot: string
): readonly [string, readonly string[]][] {
  const cached = tsconfigPathsCache.get(monorepoRoot);
  if (cached !== undefined) {
    return cached;
  }

  const tsconfigPath = path.join(monorepoRoot, 'tsconfig.base.json');
  if (!existsSync(tsconfigPath)) {
    tsconfigPathsCache.set(monorepoRoot, []);
    return [];
  }

  try {
    const parsed = JSON.parse(readFileSync(tsconfigPath, 'utf8')) as {
      readonly compilerOptions?: {
        readonly paths?: Record<string, readonly string[]>;
      };
    };
    const entries = Object.entries(parsed.compilerOptions?.paths ?? {});
    tsconfigPathsCache.set(monorepoRoot, entries);
    return entries;
  } catch {
    tsconfigPathsCache.set(monorepoRoot, []);
    return [];
  }
}

function resolveViaTsconfigPaths(
  specifier: string,
  monorepoRoot: string
): string | null {
  const pathEntries = readTsconfigPathEntries(monorepoRoot);

  for (const [pattern, targets] of pathEntries) {
    const wildcard = matchTsconfigPathPattern(pattern, specifier);
    if (wildcard === null) {
      continue;
    }

    for (const target of targets) {
      if (!target || /\.d\.[cm]?ts$/.test(target)) {
        continue;
      }

      const targetPath = applyTsconfigTargetPattern(target, wildcard);
      const candidate = path.resolve(monorepoRoot, targetPath);
      const resolved = resolveWithExtensions(candidate);
      if (resolved !== null) {
        return resolved;
      }
    }
  }

  return null;
}

function resolveLocalSourceModuleSpecifier(
  rawSpecifier: string,
  importerPath: string
): string | null {
  // Bundler asset suffixes (e.g. `./kernels.wasm?module`) are not part of the
  // on-disk path; strip them before filesystem resolution.
  const specifier = stripModuleSpecifierSuffix(rawSpecifier);
  if (!isRelativeModuleSpecifier(specifier)) {
    const monorepoRoot = findMonorepoRoot(importerPath);
    if (!monorepoRoot) {
      return null;
    }

    const tsconfigResolved = resolveViaTsconfigPaths(specifier, monorepoRoot);
    if (tsconfigResolved !== null) {
      return tsconfigResolved;
    }

    // Legacy fallback for @a0n/* aliases that are not captured in tsconfig paths.
    if (specifier.startsWith('@a0n/')) {
      const parts = specifier.split('/');
      const pkg = parts[1];
      const subpath = parts.slice(2).join('/');
      const candidates = [
        path.join(monorepoRoot, 'packages', pkg, 'src', subpath),
        path.join(monorepoRoot, 'packages', pkg, subpath),
        path.join(monorepoRoot, 'open-source', pkg, 'src', subpath),
        path.join(monorepoRoot, 'open-source', pkg, subpath),
        path.join(monorepoRoot, 'apps', pkg, 'src', subpath),
        path.join(monorepoRoot, 'apps', pkg, subpath),
        path.join(monorepoRoot, 'shared-ui', pkg, 'src', subpath),
        path.join(monorepoRoot, 'shared-ui', pkg, subpath),
        path.join(monorepoRoot, 'shared-utils', pkg, 'src', subpath),
        path.join(monorepoRoot, 'shared-utils', pkg, subpath),
      ];

      for (const base of candidates) {
        const result = resolveWithExtensions(base);
        if (result) {
          return result;
        }
      }
    }
    return null;
  }

  const resolved = path.resolve(path.dirname(importerPath), specifier);
  return resolveWithExtensions(resolved);
}

function resolveRuntimeModuleSpecifier(
  specifier: string,
  importerPath: string
): string | null {
  if (specifier.startsWith('/') || specifier.startsWith('file:')) {
    return null;
  }

  if (NODE_BUILTIN_MODULES.has(specifier) || specifier.startsWith('node:')) {
    return specifier;
  }

  try {
    const importerHref = pathToFileURL(importerPath).href;
    const resolvedHref = import.meta.resolve(specifier, importerHref);
    if (typeof resolvedHref === 'string' && resolvedHref.length > 0) {
      return resolvedHref;
    }
  } catch {
    // Fall through to CommonJS resolution for older runtimes and edge cases.
  }

  try {
    const importerRequire = createRequire(pathToFileURL(importerPath));
    const resolved = importerRequire.resolve(specifier);
    return pathToFileURL(resolved).href;
  } catch {
    return null;
  }
}

function sourceDeclaresRequire(source: string): boolean {
  return (
    /\b(?:const|let|var)\s+require\b/.test(source) ||
    /\bfunction\s+require\b/.test(source) ||
    /\bimport\s+\{\s*[^}]*\brequire\b[^}]*\}\s+from\b/.test(source) ||
    /\bimport\s+require\b/.test(source)
  );
}

function withNodeEsmPrelude(source: string, sourceFilePath: string): string {
  const prelude: string[] = [];
  let rewrittenSource = source;
  const sourceFileUrlLiteral = JSON.stringify(pathToFileURL(sourceFilePath).href);
  const sourceFilePathLiteral = JSON.stringify(sourceFilePath);
  const declaresFilename = /\b(?:const|let|var)\s+__filename\b/.test(source);
  const declaresDirname = /\b(?:const|let|var)\s+__dirname\b/.test(source);

  const needsFileContext =
    /\b__dirname\b|\b__filename\b/.test(source) ||
    /\bimport\.meta\.(dir|path)\b/.test(source);

  if (needsFileContext) {
    prelude.push("import * as __gnodePath from 'node:path';");
    if (!declaresFilename) {
      prelude.push(`const __filename = ${sourceFilePathLiteral};`);
    }
    if (!declaresDirname) {
      const dirnameSource = declaresFilename
        ? '__gnodePath.dirname(__filename)'
        : `__gnodePath.dirname(${sourceFilePathLiteral})`;
      prelude.push(`const __dirname = ${dirnameSource};`);
    }
  }

  if (/\bimport\.meta\.dir\b/.test(source)) {
    rewrittenSource = rewrittenSource.replace(
      /\bimport\.meta\.dir\b/g,
      '__dirname'
    );
  }

  if (/\bimport\.meta\.path\b/.test(source)) {
    rewrittenSource = rewrittenSource.replace(
      /\bimport\.meta\.path\b/g,
      '__filename'
    );
  }

  if (/\brequire\b/.test(source) && !sourceDeclaresRequire(source)) {
    prelude.push(
      "import { createRequire as __gnodeCreateRequire } from 'node:module';"
    );
    prelude.push(`const require = __gnodeCreateRequire(${sourceFileUrlLiteral});`);
  }

  if (prelude.length === 0) {
    return rewrittenSource;
  }

  return [...prelude, '', rewrittenSource].join('\n');
}

function rewriteImportMetaUrlToSourcePath(
  source: string,
  sourceFilePath: string
): string {
  if (!/\bimport\.meta\.url\b/.test(source)) {
    return source;
  }

  const sourceUrlLiteral = JSON.stringify(pathToFileURL(sourceFilePath).href);
  return source.replace(/\bimport\.meta\.url\b/g, sourceUrlLiteral);
}

interface ModuleSpecifierReference {
  readonly start: number;
  readonly end: number;
  readonly specifier: string;
  /** When true, emit `JSON.stringify(replacement)` (replaces an unquoted import() argument). */
  readonly jsonEncodeReplacement?: boolean;
}

function resolveImportArgumentIdentifierToSpecifier(
  sourceFile: ts.SourceFile,
  importNode: ts.CallExpression,
  identifier: ts.Identifier
): string | null {
  const importPos = importNode.getStart(sourceFile);
  let node: ts.Node | undefined = importNode;
  while (node) {
    const statements: ts.NodeArray<ts.Statement> | undefined = ts.isBlock(node)
      ? node.statements
      : ts.isSourceFile(node)
        ? node.statements
        : ts.isModuleBlock(node)
          ? node.statements
          : undefined;
    if (statements) {
      for (const stmt of statements) {
        if (stmt.getStart(sourceFile) >= importPos) {
          break;
        }
        if (ts.isVariableStatement(stmt)) {
          for (const decl of stmt.declarationList.declarations) {
            if (
              ts.isIdentifier(decl.name) &&
              decl.name.text === identifier.text &&
              decl.initializer &&
              ts.isStringLiteralLike(decl.initializer)
            ) {
              return decl.initializer.text;
            }
          }
        }
      }
    }
    node = node.parent;
  }
  return null;
}

function scriptKindForPath(filePath: string): ts.ScriptKind {
  const extension = path.extname(filePath).toLowerCase();
  if (extension === '.tsx') {
    return ts.ScriptKind.TSX;
  }
  if (extension === '.jsx') {
    return ts.ScriptKind.JSX;
  }
  if (extension === '.js' || extension === '.mjs' || extension === '.cjs') {
    return ts.ScriptKind.JS;
  }
  return ts.ScriptKind.TS;
}

function collectModuleSpecifierReferences(
  source: string,
  importerPath: string
): ModuleSpecifierReference[] {
  const sourceFile = ts.createSourceFile(
    importerPath,
    source,
    ts.ScriptTarget.Latest,
    true,
    scriptKindForPath(importerPath)
  );
  const references = new Map<number, ModuleSpecifierReference>();

  const addReference = (literal: ts.StringLiteralLike): void => {
    const start = literal.getStart(sourceFile) + 1;
    const end = literal.getEnd() - 1;
    references.set(start, {
      start,
      end,
      specifier: literal.text,
    });
  };

  const addIdentifierImportReference = (
    id: ts.Identifier,
    specifier: string
  ): void => {
    const start = id.getStart();
    const end = id.getEnd();
    references.set(start, {
      start,
      end,
      specifier,
      jsonEncodeReplacement: true,
    });
  };

  const visit = (node: ts.Node): void => {
    if (ts.isImportDeclaration(node) && ts.isStringLiteral(node.moduleSpecifier)) {
      addReference(node.moduleSpecifier);
    } else if (
      ts.isExportDeclaration(node) &&
      node.moduleSpecifier &&
      ts.isStringLiteral(node.moduleSpecifier)
    ) {
      addReference(node.moduleSpecifier);
    } else if (
      ts.isCallExpression(node) &&
      node.expression.kind === ts.SyntaxKind.ImportKeyword
    ) {
      const firstArgument = node.arguments[0];
      if (firstArgument && ts.isStringLiteralLike(firstArgument)) {
        addReference(firstArgument);
      } else if (firstArgument && ts.isIdentifier(firstArgument)) {
        const resolved = resolveImportArgumentIdentifierToSpecifier(
          sourceFile,
          node,
          firstArgument
        );
        if (resolved !== null) {
          addIdentifierImportReference(firstArgument, resolved);
        }
      }
    }

    ts.forEachChild(node, visit);
  };

  visit(sourceFile);
  return [...references.values()].sort((left, right) => right.start - left.start);
}

async function rewriteSourceImports(
  source: string,
  importerPath: string,
  shims: {
    readonly bunShimPath: string;
    readonly indexShimPath: string;
    readonly vitestShimPath: string;
  },
  moduleCache: Map<string, string>,
  createdPaths: Set<string>,
  artifactRoots: Set<string>,
  fallbackArtifactRoot: string,
  runToken: string,
  sourceToTranspiledUrl: Map<string, string>
): Promise<string> {
  const references = collectModuleSpecifierReferences(source, importerPath);
  let rewritten = source;
  for (const reference of references) {
    let replacement: string | null = null;

    const compatSpecifier = runtimeCompatSpecifier(reference.specifier, shims);
    if (compatSpecifier !== null) {
      replacement = compatSpecifier;
    } else {
      const sourceDependency = resolveLocalSourceModuleSpecifier(
        reference.specifier,
        importerPath
      );
      if (sourceDependency !== null) {
        if (isWasmAssetFile(sourceDependency)) {
          // Do NOT point the import at the raw `.wasm`: Node would ESM-instantiate
          // it and fail to resolve its host imports (e.g. `env`). Emit a shim that
          // compiles the bytes to a WebAssembly.Module (default export), matching
          // the bundler `?module` contract while keeping byte/url fallbacks.
          const shimPath = ensureWasmAssetShim(sourceDependency, createdPaths);
          replacement = canonicalFileUrl(shimPath);
        } else {
          const dependencyPath = await transpileSourceModule(
            sourceDependency,
            shims,
            moduleCache,
            createdPaths,
            false,
            artifactRoots,
            fallbackArtifactRoot,
            runToken,
            sourceToTranspiledUrl
          );
          replacement = canonicalFileUrl(dependencyPath);
        }
      } else {
        const runtimeDependency = resolveRuntimeModuleSpecifier(
          reference.specifier,
          importerPath
        );
        if (runtimeDependency !== null) {
          replacement = runtimeDependency;
        }
      }
    }

    if (replacement === null) {
      continue;
    }

    const encoded = reference.jsonEncodeReplacement
      ? JSON.stringify(replacement)
      : replacement;

    rewritten =
      rewritten.slice(0, reference.start) +
      encoded +
      rewritten.slice(reference.end);
  }

  return rewritten;
}

async function transpileSourceModule(
  filePath: string,
  shims: {
    readonly bunShimPath: string;
    readonly indexShimPath: string;
    readonly vitestShimPath: string;
  },
  moduleCache: Map<string, string>,
  createdPaths: Set<string>,
  isEntry: boolean,
  artifactRoots: Set<string>,
  fallbackArtifactRoot: string,
  runToken: string,
  sourceToTranspiledUrl: Map<string, string>
): Promise<string> {
  const cached = moduleCache.get(filePath);
  if (cached !== undefined) {
    return cached;
  }

  // Non-entry source modules are shared process-wide so the Node ESM
  // loader caches each transpiled dependency once instead of holding
  // one copy per test file. This prevents heap exhaustion at
  // package-level test runs (240+ test files).
  if (!isEntry) {
    const sharedCached = SHARED_SOURCE_MODULE_CACHE.get(filePath);
    if (sharedCached !== undefined) {
      moduleCache.set(filePath, sharedCached);
      sourceToTranspiledUrl.set(filePath, canonicalFileUrl(sharedCached));
      return sharedCached;
    }
  }

  const useSharedRoot = !isEntry;
  const roots = useSharedRoot
    ? createSharedArtifactRoots()
    : createTransientArtifactRoots(fallbackArtifactRoot, runToken);
  // The shared root is intentionally NOT added to per-file artifactRoots:
  // it persists across the whole test run so previously cached dep modules
  // remain on disk for the Node ESM loader to resolve on later test files.
  if (!useSharedRoot) {
    artifactRoots.add(roots.artifactRoot);
  }
  const hashSalt = isEntry ? `${filePath}:${runToken}` : filePath;
  const modulePath = path.join(
    roots.moduleRoot,
    `${isEntry ? '.__gnode_test__' : '.__gnode_test_dep__'}${createHash('sha1')
      .update(hashSalt)
      .digest('hex')
      .slice(0, 16)}.mjs`
  );
  moduleCache.set(filePath, modulePath);
  if (!isEntry) {
    SHARED_SOURCE_MODULE_CACHE.set(filePath, modulePath);
    SHARED_SOURCE_MODULE_PATHS.add(modulePath);
  }
  createdPaths.add(modulePath);

  const rewritten = await rewriteSourceImports(
    readFileSync(filePath, 'utf8'),
    filePath,
    shims,
    moduleCache,
    createdPaths,
    artifactRoots,
    fallbackArtifactRoot,
    runToken,
    sourceToTranspiledUrl
  );
  let transformed: Awaited<ReturnType<typeof transform>>;
  const loader = loaderForPath(filePath);
  try {
    transformed = await transform(
      rewriteImportMetaUrlToSourcePath(rewritten, filePath),
      {
        format: 'esm',
        loader,
        sourcemap: 'inline',
        sourcefile: filePath,
        target: 'node22',
      }
    );
  } catch (error) {
    const fallback = transpileWithTypeScriptFallback(
      rewriteImportMetaUrlToSourcePath(rewritten, filePath),
      filePath,
      loader
    );
    if (fallback !== null) {
      transformed = {
        ...fallback,
        code: await rewriteSourceImports(
          fallback.code,
          filePath,
          shims,
          moduleCache,
          createdPaths,
          artifactRoots,
          fallbackArtifactRoot,
          runToken,
          sourceToTranspiledUrl
        ),
      };
    } else {
      const message = error instanceof Error ? error.message : String(error);
      throw new Error(`Failed to transpile ${filePath}: ${message}`, {
        cause: error,
      });
    }
  }
  writeFileSync(modulePath, withNodeEsmPrelude(transformed.code, filePath), 'utf8');
  sourceToTranspiledUrl.set(filePath, canonicalFileUrl(modulePath));
  return modulePath;
}

function transpileWithTypeScriptFallback(
  sourceText: string,
  filePath: string,
  loader: ReturnType<typeof loaderForPath>
): Awaited<ReturnType<typeof transform>> | null {
  if (loader !== 'ts' && loader !== 'tsx') {
    return null;
  }

  const result = ts.transpileModule(sourceText, {
    fileName: filePath,
    compilerOptions: {
      target: ts.ScriptTarget.ES2022,
      module: ts.ModuleKind.ESNext,
      moduleResolution: ts.ModuleResolutionKind.Bundler,
      jsx: ts.JsxEmit.ReactJSX,
      jsxImportSource: 'react',
      sourceMap: true,
      inlineSources: true,
      esModuleInterop: true,
      isolatedModules: true,
    },
    reportDiagnostics: true,
  });
  const blockingDiagnostic = (result.diagnostics ?? []).find(
    (diagnostic) => diagnostic.category === ts.DiagnosticCategory.Error
  );
  if (blockingDiagnostic) {
    return null;
  }

  return {
    code: result.outputText,
    map: result.sourceMapText ?? '',
    warnings: [],
  };
}

function createSharedArtifactRoots(): {
  readonly artifactRoot: string;
  readonly moduleRoot: string;
  readonly shimRoot: string;
} {
  const artifactRoot = path.join(tmpdir(), 'gnode-test-runtime', '.__gnode_shared__');
  const moduleRoot = path.join(artifactRoot, 'modules');
  const shimRoot = path.join(artifactRoot, 'shims');
  mkdirSync(moduleRoot, { recursive: true });
  mkdirSync(shimRoot, { recursive: true });
  return { artifactRoot, moduleRoot, shimRoot };
}

async function transpileTypeScriptTestFile(
  filePath: string,
  packageRoot: string
): Promise<{
  readonly modulePath: string;
  readonly artifactRoots: readonly string[];
  readonly runToken: string;
  readonly sourceToTranspiledUrl: ReadonlyMap<string, string>;
}> {
  const runToken = createRunToken(filePath);
  const entryPackageRoot = findNearestPackageRoot(filePath) ?? packageRoot;
  const roots = createTransientArtifactRoots(entryPackageRoot, runToken);
  const shims = ensureRuntimeShims(roots.shimRoot);
  const createdPaths = new Set<string>();
  const artifactRoots = new Set<string>([roots.artifactRoot]);
  const moduleCache = new Map<string, string>();
  const sourceToTranspiledUrl = new Map<string, string>();
  const modulePath = await transpileSourceModule(
    filePath,
    shims,
    moduleCache,
    createdPaths,
    true,
    artifactRoots,
    packageRoot,
    runToken,
    sourceToTranspiledUrl
  );

  return {
    artifactRoots: [...artifactRoots],
    modulePath,
    runToken,
    sourceToTranspiledUrl,
  };
}

function cleanupTranspiledArtifacts(artifactRoots: readonly string[]): void {
  for (const artifactRoot of artifactRoots) {
    try {
      rmSync(artifactRoot, { recursive: true, force: true });
    } catch {
      // Best-effort cleanup; ignore transient artifact cleanup errors.
    }
  }
}

interface DomHandle {
  unregister(): Promise<void>;
}

// Register a happy-dom global environment for component tests (.tsx/.jsx), so
// document/window/navigator exist before the test graph evaluates. Returns a
// handle to tear it down, or null when no DOM is needed. happy-dom is resolved
// at runtime (external to the gnode bundle) and only loaded for DOM tests.
async function maybeRegisterDom(filePath: string): Promise<DomHandle | null> {
  if (!/\.(tsx|jsx)$/.test(filePath)) return null;
  try {
    const mod = (await import('@a0n/happy/global')) as {
      GlobalRegistrator: {
        register(options?: { url?: string }): void;
        unregister(): Promise<void>;
      };
    };
    // A concrete http origin so code using new URL(path, location.origin) works
    // (happy-dom defaults to about:blank → origin "null" → Invalid URL).
    mod.GlobalRegistrator.register({ url: 'https://storms.watch/' });
    // @testing-library/react auto-registers afterEach(cleanup) only if a GLOBAL
    // afterEach exists (it does `typeof afterEach === 'function'` at import). Our
    // hooks are module exports, not globals — expose them during DOM tests so
    // Testing Library cleans the DOM between cases (else renders accumulate →
    // "found multiple elements").
    const g = globalThis as Record<string, unknown>;
    const hadAfterEach = 'afterEach' in g;
    const hadBeforeEach = 'beforeEach' in g;
    g.afterEach = afterEach;
    g.beforeEach = beforeEach;
    return {
      unregister: async () => {
        if (!hadAfterEach) delete g.afterEach;
        if (!hadBeforeEach) delete g.beforeEach;
        await mod.GlobalRegistrator.unregister();
      },
    };
  } catch (error) {
    throw new Error(
      `Failed to set up the DOM (happy-dom) environment for ${filePath}: ${
        error instanceof Error ? error.message : String(error)
      }`
    );
  }
}

async function loadTypeScriptRegistry(
  filePath: string,
  packageRoot: string
): Promise<{
  readonly registry: GnosisTestRegistry;
  readonly cleanup: () => void;
}> {
  const registry = new GnosisTestRegistry(filePath);
  const { artifactRoots, modulePath, runToken, sourceToTranspiledUrl } =
    await transpileTypeScriptTestFile(filePath, packageRoot);
  const hooks = registerHooks({
    resolve(specifier, context, nextResolve) {
      const resolved = nextResolve(specifier, context);
      if (
        typeof resolved === 'object' &&
        resolved !== null &&
        'url' in resolved &&
        typeof resolved.url === 'string'
      ) {
        const mockUrl = resolveModuleMockRedirect(resolved.url);
        if (mockUrl !== null) {
          return {
            ...resolved,
            shortCircuit: true,
            url: mockUrl,
          };
        }
      }
      return resolved;
    },
    load(url, context, nextLoad) {
      const mockSource = loadModuleMockSource(url);
      if (mockSource !== null) {
        return {
          format: 'module',
          shortCircuit: true,
          source: mockSource,
        };
      }
      return nextLoad(url, context);
    },
  });
  setActiveModuleMockContext({
    defaultImporterPath: filePath,
    runToken,
    sourceToTranspiledUrl,
  });

  // Component tests (.tsx/.jsx) need a DOM. Register happy-dom globals
  // (document/window/...) before the test graph loads, so @testing-library and
  // React components work. Kept external from the gnode bundle + torn down after.
  const domHandle = await maybeRegisterDom(filePath);

  setActiveRegistry(registry);
  setTestRuntimeExports();
  let imported = false;
  try {
    await import(
      `${pathToFileURL(modulePath).href}?cacheBust=${encodeURIComponent(
        `${Date.now()}-${Math.random()}`
      )}`
    );
    imported = true;
  } finally {
    clearTestRuntimeExports();
    setActiveRegistry(null);
    if (!imported) {
      setActiveModuleMockContext(null);
      hooks.deregister();
      cleanupTranspiledArtifacts(artifactRoots);
      if (domHandle) void domHandle.unregister();
    }
  }

  return {
    cleanup: () => {
      setActiveModuleMockContext(null);
      hooks.deregister();
      cleanupTranspiledArtifacts(artifactRoots);
      if (domHandle) void domHandle.unregister();
    },
    registry,
  };
}

function fullCaseName(suitePath: readonly string[], caseName: string): string {
  return [...suitePath, caseName]
    .filter((segment) => segment.length > 0)
    .join(' > ');
}

function suiteLabel(suite: GnosisTestSuiteDefinition): string {
  return suite.suitePath.join(' > ') || '(root)';
}

function suiteHasSelectedCases(
  suite: GnosisTestSuiteDefinition,
  state: SuiteRunState
): boolean {
  if (state.runAllCases) {
    return true;
  }

  if (state.matcher === null) {
    return true;
  }

  return suite.items.some((item) => {
    if (item.type === 'case') {
      return state.matcher!(fullCaseName(item.suitePath, item.name));
    }
    return suiteHasSelectedCases(item, state);
  });
}

function caseMatches(
  definition: GnosisTestCaseDefinition,
  state: SuiteRunState
): boolean {
  if (state.runAllCases || state.matcher === null) {
    return true;
  }
  return state.matcher(fullCaseName(definition.suitePath, definition.name));
}

function handlerLabel(seed: string): string {
  return `TestStep${createHash('sha1')
    .update(seed)
    .digest('hex')
    .slice(0, 16)}`;
}

function buildLinearTopology(stepLabels: readonly string[]): string {
  const nodes = ['(n0:TestPassThrough)'];
  const lines: string[] = [];

  for (let index = 0; index < stepLabels.length; index += 1) {
    nodes.push(`(n${index + 1}:${stepLabels[index]})`);
  }
  nodes.push(`(n${stepLabels.length + 1}:TestPassThrough)`);

  for (let index = 0; index < nodes.length - 1; index += 1) {
    lines.push(`${nodes[index]}-[:PROCESS]->${nodes[index + 1]}`);
  }

  return `${lines.join('\n')}\n`;
}

async function runExecutionPlan(
  label: string,
  steps: readonly ExecutionStep[],
  filePath: string
): Promise<{
  readonly ok: boolean;
  readonly diagnostics: readonly string[];
  readonly formalStatus: FormalStatus;
  readonly formalReasons: readonly string[];
  readonly theoremRefs: readonly string[];
  readonly durationMs: number;
}> {
  const start = performance.now();

  if (steps.length === 0) {
    return {
      ok: true,
      diagnostics: [],
      formalStatus: 'verified',
      formalReasons: [],
      theoremRefs: [],
      durationMs: 0,
    };
  }

  const stepLabels = steps.map((step, index) =>
    handlerLabel(`${filePath}:${label}:${step.kind}:${step.name}:${index}`)
  );
  const topologySource = buildLinearTopology(stepLabels);
  const analysis = await analyzeGnosisSource(topologySource);

  if (analysis.universe.status === 'violated') {
    return {
      ok: false,
      diagnostics: analysis.universe.reasons,
      formalStatus: analysis.universe.status,
      formalReasons: analysis.universe.reasons,
      theoremRefs: analysis.universe.certificate.theoremRefs,
      durationMs: performance.now() - start,
    };
  }

  const compiler = new BettyCompiler();
  const parsed = compiler.parse(topologySource);
  const errors = parsed.diagnostics.filter(
    (diagnostic) => diagnostic.severity === 'error'
  );

  if (!parsed.ast || errors.length > 0) {
    return {
      ok: false,
      diagnostics: errors.map(
        (diagnostic) =>
          `${diagnostic.line}:${diagnostic.column} ${diagnostic.message}`
      ),
      formalStatus: 'violated',
      formalReasons: ['generated test topology failed to parse'],
      theoremRefs: [],
      durationMs: performance.now() - start,
    };
  }

  const registry = new GnosisRegistry();
  registry.register('TestPassThrough', async (payload) => payload, {
    override: true,
  });

  steps.forEach((step, index) => {
    const labelName = stepLabels[index]!;
    registry.register(
      labelName,
      async (payload) => {
        const activePayload =
          typeof payload === 'object' && payload !== null
            ? (payload as ExecutionPayload)
            : ({ failure: null } satisfies ExecutionPayload);

        if (
          activePayload.failure !== null &&
          step.kind !== 'afterEach' &&
          step.kind !== 'afterAll'
        ) {
          return activePayload;
        }

        try {
          await step.fn();
          return activePayload;
        } catch (error) {
          return {
            ...activePayload,
            failure: {
              stepName: step.name,
              error,
            },
          } satisfies ExecutionPayload;
        }
      },
      { override: true }
    );
  });

  const engine = new GnosisEngine(registry);
  const result = await engine.executeWithResult(parsed.ast, {
    failure: null,
  } satisfies ExecutionPayload);
  const payload =
    typeof result.payload === 'object' && result.payload !== null
      ? (result.payload as ExecutionPayload)
      : ({ failure: null } satisfies ExecutionPayload);

  const diagnostics =
    payload.failure === null
      ? []
      : [
          `${payload.failure.stepName}: ${
            payload.failure.error instanceof Error
              ? payload.failure.error.stack ?? payload.failure.error.message
              : String(payload.failure.error)
          }`,
        ];

  return {
    ok: payload.failure === null,
    diagnostics,
    formalStatus: result.law.status,
    formalReasons: analysis.universe.reasons,
    theoremRefs: collectTheoremRefs([
      analysis.universe.certificate.theoremRefs,
      result.law.theoremRefs,
    ]),
    durationMs: performance.now() - start,
  };
}

async function runCaseDefinition(
  definition: GnosisTestCaseDefinition,
  ancestors: readonly GnosisTestSuiteDefinition[],
  state: SuiteRunState
): Promise<GnosisTestEntryResult> {
  const beforeEachSteps = ancestors.flatMap((suite) =>
    suite.hooks.beforeEach.map(
      (fn, index): ExecutionStep => ({
        kind: 'beforeEach',
        name: `${suiteLabel(suite)} beforeEach #${index + 1}`,
        fn,
      })
    )
  );
  const afterEachSteps = [...ancestors].reverse().flatMap((suite) =>
    suite.hooks.afterEach.map(
      (fn, index): ExecutionStep => ({
        kind: 'afterEach',
        name: `${suiteLabel(suite)} afterEach #${index + 1}`,
        fn,
      })
    )
  );
  const name = fullCaseName(definition.suitePath, definition.name);
  const result = await runExecutionPlan(
    name,
    [
      ...beforeEachSteps,
      { kind: 'case', name, fn: definition.fn },
      ...afterEachSteps,
    ],
    state.filePath
  );

  return {
    kind: 'case',
    name,
    ok: result.ok,
    durationMs: result.durationMs,
    diagnostics: result.diagnostics,
    formalStatus: result.formalStatus,
    formalReasons: result.formalReasons,
    theoremRefs: result.theoremRefs,
  };
}

async function runHookSequence(
  suite: GnosisTestSuiteDefinition,
  kind: 'beforeAll' | 'afterAll',
  state: SuiteRunState
): Promise<GnosisTestEntryResult | null> {
  const hooks = suite.hooks[kind];
  if (hooks.length === 0) {
    return null;
  }

  const result = await runExecutionPlan(
    `${suiteLabel(suite)} ${kind}`,
    hooks.map(
      (fn, index): ExecutionStep => ({
        kind,
        name: `${suiteLabel(suite)} ${kind} #${index + 1}`,
        fn,
      })
    ),
    state.filePath
  );

  if (result.ok) {
    return null;
  }

  return {
    kind: 'hook',
    name: `${suiteLabel(suite)} ${kind}`,
    ok: false,
    durationMs: result.durationMs,
    diagnostics: result.diagnostics,
    formalStatus: result.formalStatus,
    formalReasons: result.formalReasons,
    theoremRefs: result.theoremRefs,
  };
}

async function runSuite(
  suite: GnosisTestSuiteDefinition,
  ancestors: readonly GnosisTestSuiteDefinition[],
  entries: GnosisTestEntryResult[],
  state: SuiteRunState
): Promise<void> {
  if (state.shouldStop) {
    return;
  }

  if (!suiteHasSelectedCases(suite, state)) {
    return;
  }

  // Include the root suite too, so file-level (top-level) beforeEach/afterEach
  // hooks run for every case — previously they were dropped for any test nested
  // in a describe(), silently skipping module-level setup like fetch stubbing.
  const activeAncestors = [...ancestors, suite];

  let canRunChildren = true;
  const beforeAllFailure =
    suite.name === '__root__'
      ? null
      : await runHookSequence(suite, 'beforeAll', state);
  if (beforeAllFailure) {
    entries.push(beforeAllFailure);
    canRunChildren = false;
    if (state.bail) {
      state.shouldStop = true;
    }
  }

  if (canRunChildren) {
    for (const item of suite.items) {
      if (state.shouldStop) {
        break;
      }

      if (item.type === 'suite') {
        await runSuite(item, activeAncestors, entries, state);
        continue;
      }

      if (!caseMatches(item, state)) {
        continue;
      }

      const caseResult = await runCaseDefinition(item, activeAncestors, state);
      entries.push(caseResult);
      if (!caseResult.ok && state.bail) {
        state.shouldStop = true;
      }
    }
  }

  const afterAllFailure =
    suite.name === '__root__'
      ? null
      : await runHookSequence(suite, 'afterAll', state);
  if (afterAllFailure) {
    entries.push(afterAllFailure);
    if (state.bail) {
      state.shouldStop = true;
    }
  }
}

/**
 * Handles the gnosis run Type Script Test File workflow.
 */
export async function runTypeScriptTestFile(
  filePath: string,
  options: Pick<GnosisTestRunOptions, 'grep' | 'bail' | 'packageRoot'> = {}
): Promise<GnosisTestFileResult | null> {
  const start = performance.now();
  const packageRoot = path.resolve(options.packageRoot ?? defaultPackageRoot());
  const matcher = buildMatcher(options.grep);
  const runAllCases = matcher !== null && matcher(normalizePath(filePath));
  let cleanup: (() => void) | null = null;

  try {
    const loaded = await loadTypeScriptRegistry(filePath, packageRoot);
    cleanup = loaded.cleanup;
    const registry = loaded.registry;
    const state: SuiteRunState = {
      filePath,
      runAllCases,
      matcher,
      bail: options.bail ?? false,
      shouldStop: false,
    };

    if (!suiteHasSelectedCases(registry.getRootSuite(), state)) {
      return null;
    }

    const entries: GnosisTestEntryResult[] = [];
    await runSuite(registry.getRootSuite(), [], entries, state);

    return {
      kind: 'ts',
      filePath,
      ok: entries.every((entry) => entry.ok),
      entries,
      durationMs: performance.now() - start,
      diagnostics: [],
      formalStatus: combineFormalStatus(
        entries.map((entry) => entry.formalStatus)
      ),
      theoremRefs: collectTheoremRefs(
        entries.map((entry) => entry.theoremRefs)
      ),
    };
  } catch (error) {
    return {
      kind: 'ts',
      filePath,
      ok: false,
      entries: [],
      durationMs: performance.now() - start,
      diagnostics: [
        error instanceof Error ? error.stack ?? error.message : String(error),
      ],
      formalStatus: 'violated',
      formalReasons: ['test module failed to load'],
      theoremRefs: [],
    };
  } finally {
    restoreAllMocks();
    if (cleanup !== null) {
      try {
        cleanup();
      } catch {
        // Best-effort cleanup for generated test artifacts.
      }
    }
  }
}

/**
 * Handles the gnosis run Wrapped Gg Test File workflow.
 */
export async function runWrappedGgTestFile(
  filePath: string
): Promise<GnosisTestFileResult> {
  const result = await runGGTestFile(filePath);
  const entries: GnosisTestEntryResult[] = result.modules.map(
    (moduleResult) => ({
      kind: 'module',
      name: moduleResult.name,
      ok: moduleResult.ok,
      durationMs: 0,
      diagnostics: moduleResult.violations,
      formalStatus: moduleResult.lawStatus,
      theoremRefs: moduleResult.theoremRefs,
      data: {
        beta1: moduleResult.beta1,
        edgeCount: moduleResult.edgeCount,
        expectedLawStatus: moduleResult.expectedLawStatus,
        forkCount: moduleResult.forkCount,
        nodeCount: moduleResult.nodeCount,
        stateCount: moduleResult.stateCount,
      },
    })
  );
  entries.push({
    kind: 'composition',
    name: 'test-topology',
    ok: result.composition.ok,
    durationMs: 0,
    diagnostics: result.composition.violations,
    formalStatus: result.composition.lawStatus,
    theoremRefs: result.composition.theoremRefs,
    data: {
      beta1: result.composition.beta1,
      edgeCount: result.composition.edgeCount,
      expectedLawStatus: result.composition.expectedLawStatus,
      forkCount: result.composition.forkCount,
      nodeCount: result.composition.nodeCount,
      stateCount: result.composition.stateCount,
    },
  });

  return {
    kind: 'gg',
    filePath,
    ok: result.ok,
    entries,
    durationMs: result.elapsed,
    diagnostics: [],
    formalStatus: combineFormalStatus(
      entries.map((entry) => entry.formalStatus)
    ),
    theoremRefs: collectTheoremRefs(entries.map((entry) => entry.theoremRefs)),
  };
}

async function runDiscoveredFile(
  discovered: DiscoveredTestFile,
  options: GnosisTestRunOptions
): Promise<GnosisTestFileResult | null> {
  if (discovered.kind === 'gg') {
    const matcher = buildMatcher(options.grep);
    if (matcher && !matcher(normalizePath(discovered.filePath))) {
      return null;
    }
    return runWrappedGgTestFile(discovered.filePath);
  }

  return runTypeScriptTestFile(discovered.filePath, options);
}

/**
 * Handles the gnosis run Gnosis Tests workflow.
 */
export async function runGnosisTests(
  options: GnosisTestRunOptions = {}
): Promise<GnosisTestRunResult> {
  const start = performance.now();
  const packageRoot = path.resolve(options.packageRoot ?? defaultPackageRoot());
  const discovered = discoverGnosisTestFiles(options);
  const laneCount = Math.max(1, options.lanes ?? 1);
  const fileOrder = new Map(
    discovered.map((file) => [file.filePath, file.index])
  );
  const results: GnosisTestFileResult[] = [];
  const ggRuntimeHistory = loadGgRuntimeHistory(packageRoot);
  let shouldStop = false;

  const typeScriptFiles = discovered.filter((file) => file.kind === 'ts');
  const ggMatcher = buildMatcher(options.grep);
  const ggFiles = discovered.filter(
    (file) =>
      file.kind === 'gg' &&
      (ggMatcher === null || ggMatcher(normalizePath(file.filePath)))
  );

  for (const discoveredFile of typeScriptFiles) {
    const result = await runDiscoveredFile(discoveredFile, options);
    if (result) {
      results.push(result);
      if (!result.ok && options.bail) {
        shouldStop = true;
        break;
      }
    }
  }

  if (!shouldStop) {
    if ((options.strategy ?? 'cannon') === 'linear' || laneCount === 1) {
      for (const discoveredFile of ggFiles) {
        const result = await runDiscoveredFile(discoveredFile, options);
        if (result) {
          results.push(result);
          ggRuntimeHistory.set(
            result.filePath,
            updateGgRuntimeHistoryEntry(
              ggRuntimeHistory.get(result.filePath),
              result.durationMs
            )
          );
          if (!result.ok && options.bail) {
            shouldStop = true;
            break;
          }
        }
      }
    } else {
      const ggByPath = new Map(ggFiles.map((file) => [file.filePath, file]));
      const rotation = planGgTestLanes(
        ggFiles.map((file) => file.filePath),
        {
          laneCount,
          packageRoot,
        }
      );
      await Promise.all(
        rotation.lanes.map(async (lane) => {
          for (const filePath of lane.items) {
            if (shouldStop) {
              break;
            }

            const discoveredFile = ggByPath.get(filePath);
            if (!discoveredFile) {
              continue;
            }
            const result = await runDiscoveredFile(discoveredFile, options);
            if (result) {
              results.push(result);
              ggRuntimeHistory.set(
                result.filePath,
                updateGgRuntimeHistoryEntry(
                  ggRuntimeHistory.get(result.filePath),
                  result.durationMs
                )
              );
              if (!result.ok && options.bail) {
                shouldStop = true;
              }
            }
          }
        })
      );
    }
  }

  if (!shouldStop && options.includePact) {
    const pactResult = await runPactVerification(options);
    if (pactResult.file) {
      results.push(pactResult.file);
      if (!pactResult.file.ok && options.bail) {
        shouldStop = true;
      }
    }
  }

  if (ggFiles.length > 0) {
    saveGgRuntimeHistory(packageRoot, ggRuntimeHistory);
  }

  const files = [...results].sort((left, right) => {
    const leftIndex = fileOrder.get(left.filePath) ?? Number.MAX_SAFE_INTEGER;
    const rightIndex = fileOrder.get(right.filePath) ?? Number.MAX_SAFE_INTEGER;
    if (leftIndex !== rightIndex) {
      return leftIndex - rightIndex;
    }
    return left.filePath.localeCompare(right.filePath);
  });
  const failedCount = files.filter((file) => !file.ok).length;

  return {
    ok: failedCount === 0,
    files,
    durationMs: performance.now() - start,
    fileCount: files.length,
    passedCount: files.length - failedCount,
    failedCount,
  };
}

/**
 * Formats the Gnosis Test Run.
 */
export function formatGnosisTestRun(result: GnosisTestRunResult): string {
  const lines: string[] = [];

  for (const file of result.files) {
    lines.push(`[gnode test] ${file.kind.toUpperCase()} ${file.filePath}`);
    if (file.formalStatus) {
      lines.push(`  law=${file.formalStatus}`);
    }
    if (file.theoremRefs && file.theoremRefs.length > 0) {
      lines.push(`  theorems: ${file.theoremRefs.join(', ')}`);
    }
    if (file.diagnostics.length > 0) {
      for (const diagnostic of file.diagnostics) {
        lines.push(`  ${diagnostic}`);
      }
    }

    for (const entry of file.entries) {
      lines.push(
        `  ${entry.ok ? 'PASS' : 'FAIL'} ${entry.kind} ${entry.name}${
          entry.formalStatus ? ` law=${entry.formalStatus}` : ''
        }`
      );
      if (entry.theoremRefs && entry.theoremRefs.length > 0) {
        lines.push(`         theorems: ${entry.theoremRefs.join(', ')}`);
      }
      for (const diagnostic of entry.diagnostics) {
        lines.push(`         ${diagnostic}`);
      }
    }

    lines.push(
      `  ${file.ok ? 'OK' : 'FAILED'} ${file.durationMs.toFixed(0)}ms`
    );
    lines.push('');
  }

  lines.push(
    `${result.ok ? 'ALL PASSED' : 'FAILURES'} ${result.passedCount}/${
      result.fileCount
    } files ${result.durationMs.toFixed(0)}ms`
  );

  return lines.join('\n');
}
