// @a0n/.ignored_gnosis/src/testing/types.ts
export type GnosisTestFn = () => unknown | Promise<unknown>;

export type GnosisHookKind =
  | 'beforeAll'
  | 'afterAll'
  | 'beforeEach'
  | 'afterEach';

export interface GnosisTestCaseDefinition {
  readonly type: 'case';
  readonly name: string;
  readonly fn: GnosisTestFn;
  readonly suitePath: readonly string[];
  readonly declaredAt: number;
}

export interface GnosisTestSuiteDefinition {
  readonly type: 'suite';
  readonly name: string;
  readonly suitePath: readonly string[];
  readonly declaredAt: number;
  readonly items: Array<GnosisTestSuiteDefinition | GnosisTestCaseDefinition>;
  readonly hooks: Record<GnosisHookKind, GnosisTestFn[]>;
}

export type GnosisTestEntryKind =
  | 'case'
  | 'hook'
  | 'module'
  | 'composition'
  | 'contract';

export interface GnosisTestEntryResult {
  readonly kind: GnosisTestEntryKind;
  readonly name: string;
  readonly ok: boolean;
  readonly durationMs: number;
  readonly diagnostics: readonly string[];
  readonly formalStatus?: 'verified' | 'open' | 'violated';
  readonly formalReasons?: readonly string[];
  readonly theoremRefs?: readonly string[];
  readonly data?: unknown;
}

export interface GnosisTestFileResult {
  readonly kind: 'ts' | 'gg' | 'pact';
  readonly filePath: string;
  readonly ok: boolean;
  readonly entries: readonly GnosisTestEntryResult[];
  readonly durationMs: number;
  readonly diagnostics: readonly string[];
  readonly formalStatus?: 'verified' | 'open' | 'violated';
  readonly formalReasons?: readonly string[];
  readonly theoremRefs?: readonly string[];
}

export interface GnosisTestRunResult {
  readonly ok: boolean;
  readonly files: readonly GnosisTestFileResult[];
  readonly durationMs: number;
  readonly fileCount: number;
  readonly passedCount: number;
  readonly failedCount: number;
}

export interface GnosisTestRunOptions {
  readonly targets?: readonly string[];
  readonly roots?: readonly string[];
  readonly grep?: string;
  readonly bail?: boolean;
  readonly lanes?: number;
  readonly strategy?: 'cannon' | 'linear';
  readonly packageRoot?: string;
  readonly workspaceRoot?: string;
  readonly includePact?: boolean;
  readonly pactScan?: boolean;
  readonly pactUpdateDocs?: boolean;
  readonly sarif?: boolean;
}

export class UnsupportedGnosisTestFeatureError extends Error {
  public readonly feature: string;

  public constructor(feature: string) {
    super(
      `Unsupported Gnosis test feature: ${feature}. Rewrite the suite to use the canonical @a0n/gnosis/test surface.`
    );
    this.name = 'UnsupportedGnosisTestFeatureError';
    this.feature = feature;
  }
}

/**
 * Handles the gnosis unsupported Feature workflow.
 */
export function unsupportedFeature(feature: string): never {
  throw new UnsupportedGnosisTestFeatureError(feature);
}
