// @a0n/gnosis/src/testing/expect.ts
import { strict as assert } from 'node:assert';
import { inspect } from 'node:util';
import { isMockFunction } from './mock';

type Constructor<T = unknown> = new (...args: any[]) => T;

function formatValue(value: unknown): string {
  return inspect(value, {
    depth: 6,
    breakLength: 80,
    compact: false,
  });
}

function assertMatch(pass: boolean, negated: boolean, message: string): void {
  if (negated ? pass : !pass) {
    assert.fail(message);
  }
}

function isPromiseLike(value: unknown): value is PromiseLike<unknown> {
  return (
    typeof value === 'object' &&
    value !== null &&
    'then' in value &&
    typeof (value as PromiseLike<unknown>).then === 'function'
  );
}

function extractErrorMessage(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

function partialDeepMatch(actual: unknown, expected: unknown): boolean {
  if (
    expected === null ||
    typeof expected !== 'object' ||
    actual === null ||
    typeof actual !== 'object'
  ) {
    try {
      assert.deepStrictEqual(actual, expected);
      return true;
    } catch {
      return false;
    }
  }

  if (Array.isArray(expected)) {
    if (!Array.isArray(actual) || actual.length < expected.length) {
      return false;
    }
    return expected.every((item, index) =>
      partialDeepMatch(actual[index], item)
    );
  }

  const actualRecord = actual as Record<string, unknown>;
  return Object.entries(expected as Record<string, unknown>).every(
    ([key, value]) =>
      key in actualRecord && partialDeepMatch(actualRecord[key], value)
  );
}

type PropertyPathSegment = string | number | symbol;

type PropertyPath = string | readonly PropertyPathSegment[];

function normalizePropertyPath(path: PropertyPath): readonly PropertyPathSegment[] {
  if (typeof path !== 'string') {
    return path;
  }
  const segments: PropertyPathSegment[] = path
    .split('.')
    .filter((segment: string) => segment.length > 0)
    .map((segment: string): PropertyPathSegment =>
      /^\d+$/.test(segment) ? Number.parseInt(segment, 10) : segment
    );
  return segments;
}

function resolvePropertyPath(
  value: unknown,
  path: PropertyPath
): { exists: boolean; value: unknown } {
  const segments = normalizePropertyPath(path);
  if (segments.length === 0) {
    return { exists: false, value: undefined };
  }

  let current: unknown = value;
  for (const segment of segments) {
    if (
      current === null ||
      (typeof current !== 'object' && typeof current !== 'function')
    ) {
      return { exists: false, value: undefined };
    }

    const record = current as Record<string | number | symbol, unknown>;
    if (!(segment in record)) {
      return { exists: false, value: undefined };
    }
    current = record[segment];
  }

  return { exists: true, value: current };
}

function assertThrown(
  thrown: unknown,
  expected?: string | RegExp | Constructor,
  negated = false
): void {
  const message = extractErrorMessage(thrown);

  if (expected === undefined) {
    assertMatch(
      true,
      negated,
      `expected value ${negated ? 'not ' : ''}to throw`
    );
    return;
  }

  let pass = false;
  if (typeof expected === 'string') {
    pass = message.includes(expected);
  } else if (expected instanceof RegExp) {
    pass = expected.test(message);
  } else {
    pass = thrown instanceof expected;
  }

  assertMatch(
    pass,
    negated,
    `expected thrown value ${negated ? 'not ' : ''}to match ${formatValue(
      expected
    )}, received ${formatValue(thrown)}`
  );
}

class AsyncExpectation {
  private readonly actual: PromiseLike<unknown>;
  private readonly negated: boolean;
  private readonly mode: 'resolves' | 'rejects';

  public constructor(
    actual: PromiseLike<unknown>,
    negated: boolean,
    mode: 'resolves' | 'rejects'
  ) {
    this.actual = actual;
    this.negated = negated;
    this.mode = mode;
  }

  public get not(): AsyncExpectation {
    return new AsyncExpectation(this.actual, !this.negated, this.mode);
  }

  private async unwrap(): Promise<unknown> {
    try {
      const value = await this.actual;
      if (this.mode === 'rejects') {
        assert.fail('expected promise to reject');
      }
      return value;
    } catch (error) {
      if (this.mode === 'resolves') {
        throw error;
      }
      return error;
    }
  }

  public async toBe(expected: unknown): Promise<void> {
    new SyncExpectation(await this.unwrap(), this.negated).toBe(expected);
  }

  public async toEqual(expected: unknown): Promise<void> {
    new SyncExpectation(await this.unwrap(), this.negated).toEqual(expected);
  }

  public async toMatchObject(expected: unknown): Promise<void> {
    new SyncExpectation(await this.unwrap(), this.negated).toMatchObject(
      expected
    );
  }

  public async toThrow(
    expected?: string | RegExp | Constructor
  ): Promise<void> {
    if (this.mode === 'rejects') {
      assertThrown(await this.unwrap(), expected, this.negated);
      return;
    }

    new SyncExpectation(await this.unwrap(), this.negated).toThrow(expected);
  }

  public async toBeInTheDocument(): Promise<void> {
    new SyncExpectation(await this.unwrap(), this.negated).toBeInTheDocument();
  }
}

export class SyncExpectation {
  private readonly actual: unknown;
  private readonly negated: boolean;

  public constructor(
    actual: unknown,
    negated: boolean
  ) {
    this.actual = actual;
    this.negated = negated;
  }

  public get not(): SyncExpectation {
    return new SyncExpectation(this.actual, !this.negated);
  }

  public get resolves(): AsyncExpectation {
    return new AsyncExpectation(
      Promise.resolve(this.actual),
      this.negated,
      'resolves'
    );
  }

  public get rejects(): AsyncExpectation {
    return new AsyncExpectation(
      Promise.resolve(this.actual),
      this.negated,
      'rejects'
    );
  }

  public toBe(expected: unknown): void {
    if (this.negated) {
      assert.notStrictEqual(this.actual, expected);
      return;
    }
    assert.strictEqual(this.actual, expected);
  }

  public toEqual(expected: unknown): void {
    if (this.negated) {
      assert.notDeepStrictEqual(this.actual, expected);
      return;
    }
    assert.deepStrictEqual(this.actual, expected);
  }

  public toContain(expected: unknown): void {
    const value = this.actual as string | readonly unknown[];
    const pass =
      typeof value === 'string'
        ? value.includes(String(expected))
        : Array.isArray(value)
        ? value.includes(expected)
        : false;
    assertMatch(
      pass,
      this.negated,
      `expected ${formatValue(this.actual)} ${
        this.negated ? 'not ' : ''
      }to contain ${formatValue(expected)}`
    );
  }

  // jest-dom-compatible DOM matcher (used by @testing-library suites).
  public toBeInTheDocument(): void {
    const node = this.actual as Node | null;
    const inDoc =
      node != null &&
      typeof node === 'object' &&
      typeof (node as Node).nodeType === 'number' &&
      ((node as Node).isConnected === true ||
        (typeof document !== 'undefined' &&
          typeof document.contains === 'function' &&
          document.contains(node as Node)));
    assertMatch(
      inDoc,
      this.negated,
      `expected element ${this.negated ? 'not ' : ''}to be in the document`
    );
  }

  public toMatch(expected: RegExp | string): void {
    if (typeof this.actual !== 'string') {
      assert.fail(
        `toMatch expects a string, received ${formatValue(this.actual)}`
      );
    }
    const pass =
      typeof expected === 'string'
        ? this.actual.includes(expected)
        : expected.test(this.actual);
    assertMatch(
      pass,
      this.negated,
      `expected ${formatValue(this.actual)} ${
        this.negated ? 'not ' : ''
      }to match ${formatValue(expected)}`
    );
  }

  public toMatchObject(expected: unknown): void {
    const pass = partialDeepMatch(this.actual, expected);
    assertMatch(
      pass,
      this.negated,
      `expected ${formatValue(this.actual)} ${
        this.negated ? 'not ' : ''
      }to match object ${formatValue(expected)}`
    );
  }

  public toHaveLength(expected: number): void {
    const value = this.actual as { length?: unknown };
    if (typeof value?.length !== 'number') {
      assert.fail(
        `toHaveLength expects a length-bearing value, received ${formatValue(
          this.actual
        )}`
      );
    }
    assertMatch(
      value.length === expected,
      this.negated,
      `expected length ${
        this.negated ? 'not ' : ''
      }to be ${expected}, received ${value.length}`
    );
  }

  public toHaveProperty(path: PropertyPath, expected?: unknown): void {
    const resolved = resolvePropertyPath(this.actual, path);
    const hasExpectedValue =
      arguments.length < 2 || partialDeepMatch(resolved.value, expected);
    const pass = resolved.exists && hasExpectedValue;
    assertMatch(
      pass,
      this.negated,
      `expected ${formatValue(this.actual)} ${
        this.negated ? 'not ' : ''
      }to have property ${formatValue(path)}${
        arguments.length < 2 ? '' : ` with value ${formatValue(expected)}`
      }`
    );
  }

  public toBeDefined(): void {
    assertMatch(
      typeof this.actual !== 'undefined',
      this.negated,
      `expected ${this.negated ? 'not ' : ''}to be defined`
    );
  }

  public toBeUndefined(): void {
    assertMatch(
      typeof this.actual === 'undefined',
      this.negated,
      `expected ${this.negated ? 'not ' : ''}to be undefined`
    );
  }

  public toBeNull(): void {
    assertMatch(
      this.actual === null,
      this.negated,
      `expected ${this.negated ? 'not ' : ''}to be null`
    );
  }

  public toBeTruthy(): void {
    assertMatch(
      Boolean(this.actual),
      this.negated,
      `expected ${formatValue(this.actual)} ${
        this.negated ? 'not ' : ''
      }to be truthy`
    );
  }

  public toBeNaN(): void {
    if (typeof this.actual !== 'number') {
      assert.fail(`toBeNaN expects a number, received ${formatValue(this.actual)}`);
    }
    assertMatch(
      Number.isNaN(this.actual),
      this.negated,
      `expected ${formatValue(this.actual)} ${
        this.negated ? 'not ' : ''
      }to be NaN`
    );
  }

  public toBeInstanceOf(expected: Constructor): void {
    assertMatch(
      this.actual instanceof expected,
      this.negated,
      `expected ${formatValue(this.actual)} ${
        this.negated ? 'not ' : ''
      }to be an instance of ${expected.name}`
    );
  }

  public toBeCloseTo(expected: number, precision = 2): void {
    if (typeof this.actual !== 'number') {
      assert.fail(
        `toBeCloseTo expects a number, received ${formatValue(this.actual)}`
      );
    }
    const diff = Math.abs(this.actual - expected);
    const epsilon = Math.pow(10, -precision) / 2;
    assertMatch(
      diff < epsilon,
      this.negated,
      `expected ${this.actual} ${
        this.negated ? 'not ' : ''
      }to be close to ${expected} with precision ${precision}`
    );
  }

  public toBeGreaterThan(expected: number): void {
    if (typeof this.actual !== 'number') {
      assert.fail(
        `toBeGreaterThan expects a number, received ${formatValue(this.actual)}`
      );
    }
    assertMatch(
      this.actual > expected,
      this.negated,
      `expected ${this.actual} ${
        this.negated ? 'not ' : ''
      }to be greater than ${expected}`
    );
  }

  public toBeGreaterThanOrEqual(expected: number): void {
    if (typeof this.actual !== 'number') {
      assert.fail(
        `toBeGreaterThanOrEqual expects a number, received ${formatValue(
          this.actual
        )}`
      );
    }
    assertMatch(
      this.actual >= expected,
      this.negated,
      `expected ${this.actual} ${
        this.negated ? 'not ' : ''
      }to be greater than or equal to ${expected}`
    );
  }

  public toBeLessThan(expected: number): void {
    if (typeof this.actual !== 'number') {
      assert.fail(
        `toBeLessThan expects a number, received ${formatValue(this.actual)}`
      );
    }
    assertMatch(
      this.actual < expected,
      this.negated,
      `expected ${this.actual} ${
        this.negated ? 'not ' : ''
      }to be less than ${expected}`
    );
  }

  public toBeLessThanOrEqual(expected: number): void {
    if (typeof this.actual !== 'number') {
      assert.fail(
        `toBeLessThanOrEqual expects a number, received ${formatValue(
          this.actual
        )}`
      );
    }
    assertMatch(
      this.actual <= expected,
      this.negated,
      `expected ${this.actual} ${
        this.negated ? 'not ' : ''
      }to be less than or equal to ${expected}`
    );
  }

  public toThrow(expected?: string | RegExp | Constructor): void {
    if (typeof this.actual !== 'function') {
      assert.fail(
        `toThrow expects a function, received ${formatValue(this.actual)}`
      );
    }

    let thrown = false;
    let thrownValue: unknown;
    try {
      (this.actual as () => unknown)();
    } catch (error) {
      thrown = true;
      thrownValue = error;
    }

    if (!thrown) {
      assertMatch(false, this.negated, 'expected function to throw');
      return;
    }

    assertThrown(thrownValue, expected, this.negated);
  }

  public toHaveBeenCalled(): void {
    if (!isMockFunction(this.actual)) {
      assert.fail(
        `toHaveBeenCalled expects a mock function, received ${formatValue(
          this.actual
        )}`
      );
    }
    assertMatch(
      this.actual.mock.calls.length > 0,
      this.negated,
      `expected mock ${this.negated ? 'not ' : ''}to have been called`
    );
  }

  public toHaveBeenCalledTimes(expected: number): void {
    if (!isMockFunction(this.actual)) {
      assert.fail(`toHaveBeenCalledTimes expects a mock function, received ${formatValue(this.actual)}`);
    }
    assertMatch(
      this.actual.mock.calls.length === expected,
      this.negated,
      `expected mock ${this.negated ? 'not ' : ''}to have been called ${expected} times, received ${this.actual.mock.calls.length}`
    );
  }

  public toHaveBeenCalledWith(...args: any[]): void {
    if (!isMockFunction(this.actual)) {
      assert.fail(`toHaveBeenCalledWith expects a mock function, received ${formatValue(this.actual)}`);
    }
    const pass = this.actual.mock.calls.some((call) => partialDeepMatch(call, args));
    assertMatch(
      pass,
      this.negated,
      `expected mock ${this.negated ? 'not ' : ''}to have been called with ${formatValue(args)}`
    );
  }

  public toHaveBeenLastCalledWith(...args: any[]): void {
    if (!isMockFunction(this.actual)) {
      assert.fail(`toHaveBeenLastCalledWith expects a mock function, received ${formatValue(this.actual)}`);
    }
    const lastCall = this.actual.mock.calls[this.actual.mock.calls.length - 1];
    const pass = lastCall !== undefined && partialDeepMatch(lastCall, args);
    assertMatch(
      pass,
      this.negated,
      `expected mock ${this.negated ? 'not ' : ''}to have been last called with ${formatValue(args)}`
    );
  }
}

export interface GnosisExpect {
  (actual: unknown): SyncExpectation;
  objectContaining(expected: unknown): any;
}

export const expect: GnosisExpect = ((actual: unknown) => {
  return new SyncExpectation(actual, false);
}) as GnosisExpect;

expect.objectContaining = (expected: unknown) => {
  return {
    asMatcher: () => (actual: unknown) => partialDeepMatch(actual, expected),
    // For type safety in tests that use expect.objectContaining
    toString: () => `objectContaining(${formatValue(expected)})`,
  };
};
