// @a0n/gnosis/src/testing/fixture.ts
import { existsSync, readFileSync, readdirSync } from 'node:fs';
import path from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';

const EXCLUDED_DIRS = new Set([
  '.git',
  '.tmp-test-fixtures',
  '.tmp-test-workspaces',
  'coverage',
  'dist',
  'node_modules',
]);

const fixtureSearchCache = new Map<string, readonly string[]>();

type FixtureSource = string | URL;

export interface ResolveGnosisFixtureOptions {
  readonly packageRoot?: string;
  readonly searchRoots?: readonly string[];
}

function toPath(source: FixtureSource): string {
  if (source instanceof URL) {
    return fileURLToPath(source);
  }

  if (source.startsWith('file:')) {
    return fileURLToPath(source);
  }

  return source;
}

function walkUpToPackageRoot(startPath: string): string {
  let current = path.dirname(startPath);

  while (current !== path.dirname(current)) {
    if (
      existsSync(path.join(current, 'package.json')) &&
      existsSync(path.join(current, 'gnosis.mod'))
    ) {
      return current;
    }
    current = path.dirname(current);
  }

  throw new Error(
    `Unable to locate the @a0n/gnosis package root from ${startPath}`
  );
}

function collectByBasename(root: string, basename: string): readonly string[] {
  const cacheKey = `${root}:${basename}`;
  const cached = fixtureSearchCache.get(cacheKey);
  if (cached !== undefined) {
    return cached;
  }

  const matches: string[] = [];

  if (!existsSync(root)) {
    fixtureSearchCache.set(cacheKey, matches);
    return matches;
  }

  const visit = (directory: string): void => {
    for (const entry of readdirSync(directory, { withFileTypes: true })) {
      if (entry.isDirectory()) {
        if (EXCLUDED_DIRS.has(entry.name)) {
          continue;
        }
        visit(path.join(directory, entry.name));
        continue;
      }

      if (entry.name === basename) {
        matches.push(path.join(directory, entry.name));
      }
    }
  };

  visit(root);
  fixtureSearchCache.set(cacheKey, matches);
  return matches;
}

/**
 * Resolves the Gnosis Fixture Path.
 */
export function resolveGnosisFixturePath(
  specifier: string,
  from: FixtureSource,
  options: ResolveGnosisFixtureOptions = {}
): string {
  const fromPath = toPath(from);
  if (path.isAbsolute(specifier)) {
    return specifier;
  }

  const directPath = path.resolve(path.dirname(fromPath), specifier);
  if (existsSync(directPath)) {
    return directPath;
  }

  const packageRoot =
    options.packageRoot !== undefined
      ? path.resolve(options.packageRoot)
      : walkUpToPackageRoot(fromPath);
  const searchRoots = options.searchRoots?.map((root) =>
    path.resolve(packageRoot, root)
  ) ?? [
    packageRoot,
    path.join(packageRoot, 'topologies'),
    path.join(packageRoot, 'examples'),
    path.join(packageRoot, 'lean'),
    path.join(packageRoot, 'tla'),
  ];
  const basename = path.basename(specifier);
  const matches = [
    ...new Set(
      searchRoots.flatMap((root) => collectByBasename(root, basename))
    ),
  ];

  if (matches.length === 1) {
    return matches[0]!;
  }

  if (matches.length > 1) {
    throw new Error(
      [
        `Ambiguous Gnosis fixture "${specifier}" from ${fromPath}.`,
        `Direct path ${directPath} does not exist.`,
        'Matches:',
        ...matches.map((match) => `- ${match}`),
      ].join('\n')
    );
  }

  throw new Error(
    [
      `Unable to resolve Gnosis fixture "${specifier}" from ${fromPath}.`,
      `Direct path ${directPath} does not exist.`,
      `Searched roots: ${searchRoots.join(', ')}`,
    ].join('\n')
  );
}

/**
 * Resolves the Gnosis Fixture Url.
 */
export function resolveGnosisFixtureUrl(
  specifier: string,
  from: FixtureSource,
  options: ResolveGnosisFixtureOptions = {}
): URL {
  return pathToFileURL(resolveGnosisFixturePath(specifier, from, options));
}

/**
 * Handles the gnosis read Gnosis Fixture workflow.
 */
export function readGnosisFixture(
  specifier: string,
  from: FixtureSource,
  options: ResolveGnosisFixtureOptions = {}
): string {
  return readFileSync(
    resolveGnosisFixturePath(specifier, from, options),
    'utf8'
  );
}
