// @a0n/gnosis/src/testing/mutation-report.ts
/**
 * Stryker Mutation Testing Elements report JSON
 * (mutation-testing-report schema — not SARIF).
 *
 * @see https://stryker-mutator.io/blog/one-mutation-testing-html-report/
 * @see https://github.com/stryker-mutator/mutation-testing-elements/blob/master/packages/report-schema/src/mutation-testing-report-schema.json
 */

import * as path from 'node:path';

/** MutantStatus from mutation-testing-report-schema.json */
export type MutantStatus =
  | 'Killed'
  | 'Survived'
  | 'NoCoverage'
  | 'CompileError'
  | 'RuntimeError'
  | 'Timeout'
  | 'Ignored'
  | 'Pending';

export interface MutationPosition {
  readonly line: number;
  readonly column: number;
}

export interface MutationLocation {
  readonly start: MutationPosition;
  readonly end: MutationPosition;
}

export interface MutantResult {
  readonly id: string;
  readonly mutatorName: string;
  readonly location: MutationLocation;
  readonly status: MutantStatus;
  readonly replacement?: string;
  readonly description?: string;
  readonly statusReason?: string;
  readonly static?: boolean;
}

export interface FileResult {
  readonly language: string;
  readonly source: string;
  readonly mutants: readonly MutantResult[];
}

export interface MutationTestingReport {
  readonly schemaVersion: string;
  readonly thresholds: { readonly high: number; readonly low: number };
  readonly files: Record<string, FileResult>;
  readonly projectRoot?: string;
  readonly framework?: {
    readonly name: string;
    readonly version?: string;
    readonly branding?: {
      readonly homepageUrl: string;
      readonly imageUrl?: string;
    };
  };
  readonly config?: Record<string, unknown>;
}

export interface CompilerDiagnosticLike {
  readonly line: number;
  readonly column: number;
  readonly code?: string;
  readonly message: string;
  readonly severity: 'error' | 'warning' | 'info';
}

export interface CorrectnessViolationLike {
  readonly kind: string;
  readonly name: string;
  readonly message: string;
}

export interface CapabilityIssueLike {
  readonly severity: 'error' | 'warning' | 'info';
  readonly capability: string;
  readonly message: string;
}

export interface LeanCheckLike {
  readonly attempted: boolean;
  readonly ok: boolean;
  readonly message: string;
}

function locationFromLineColumn(line: number, column: number): MutationLocation {
  const L = Math.max(1, line);
  const C = Math.max(1, column);
  return {
    start: { line: L, column: C },
    end: { line: L, column: C + 1 },
  };
}

function reportFileKey(absolutePath: string, cwd: string): string {
  const rel = path.relative(cwd, absolutePath);
  if (rel.startsWith(`..${path.sep}`) || rel === '..') {
    return path.basename(absolutePath);
  }
  return rel.split(path.sep).join('/');
}

function diagnosticToStatus(
  severity: CompilerDiagnosticLike['severity']
): MutantStatus {
  if (severity === 'error') {
    return 'CompileError';
  }
  if (severity === 'warning') {
    return 'RuntimeError';
  }
  return 'Ignored';
}

export interface BuildGnosisGgMutationReportInput {
  readonly filePath: string;
  readonly source: string;
  readonly cwd: string;
  readonly mode: string;
  readonly compilerDiagnostics: readonly CompilerDiagnosticLike[];
  readonly correctnessViolations: readonly CorrectnessViolationLike[];
  readonly capabilityErrors: readonly CapabilityIssueLike[];
  readonly universeOk: boolean;
  readonly universeReasons: readonly string[];
  readonly buleyExceeded: boolean;
  readonly buleyNumber: number;
  readonly maxBuley: number | null;
  readonly leanCheck: LeanCheckLike | null;
  readonly gnosisVersion?: string;
  readonly thresholds?: { high: number; low: number };
}

/**
 * Maps Gnosis `.gg` / `verify` / `lint` findings into mutation-testing-elements JSON.
 * Each finding is a “mutant” of the source under test; static gates that reject it
 * use {@link MutantStatus} values such as `CompileError` or `Killed`.
 */
export function buildGnosisGgMutationReport(
  input: BuildGnosisGgMutationReportInput
): MutationTestingReport {
  const thresholds = input.thresholds ?? { high: 80, low: 60 };
  const key = reportFileKey(input.filePath, input.cwd);
  const mutants: MutantResult[] = [];
  let id = 0;
  const nextId = (): string => {
    id += 1;
    return `gnosis-gg-${id}`;
  };

  for (const d of input.compilerDiagnostics) {
    mutants.push({
      id: nextId(),
      mutatorName: d.code ?? 'GnosisCompilerDiagnostic',
      location: locationFromLineColumn(d.line, d.column),
      status: diagnosticToStatus(d.severity),
      description: d.message,
      statusReason: d.message,
    });
  }

  for (const v of input.correctnessViolations) {
    mutants.push({
      id: nextId(),
      mutatorName: `Correctness:${v.kind}`,
      location: locationFromLineColumn(1, 1),
      status: 'Killed',
      description: `${v.name}: ${v.message}`,
      statusReason: `${v.kind}:${v.name} — ${v.message}`,
    });
  }

  for (const issue of input.capabilityErrors) {
    if (issue.severity !== 'error') {
      continue;
    }
    mutants.push({
      id: nextId(),
      mutatorName: `Capability:${issue.capability}`,
      location: locationFromLineColumn(1, 1),
      status: 'CompileError',
      description: issue.message,
      statusReason: issue.message,
    });
  }

  if (!input.universeOk) {
    for (const reason of input.universeReasons) {
      mutants.push({
        id: nextId(),
        mutatorName: 'UniverseAdmission',
        location: locationFromLineColumn(1, 1),
        status: 'Killed',
        description: reason,
        statusReason: reason,
      });
    }
  }

  if (input.buleyExceeded && input.maxBuley !== null) {
    mutants.push({
      id: nextId(),
      mutatorName: 'BuleyCeiling',
      location: locationFromLineColumn(1, 1),
      status: 'RuntimeError',
      description: `Buley ${input.buleyNumber} exceeds max ${input.maxBuley}`,
      statusReason: `buley ${input.buleyNumber} > ${input.maxBuley}`,
    });
  }

  if (input.leanCheck?.attempted && !input.leanCheck.ok) {
    mutants.push({
      id: nextId(),
      mutatorName: 'LeanProofCheck',
      location: locationFromLineColumn(1, 1),
      status: 'CompileError',
      description: input.leanCheck.message,
      statusReason: input.leanCheck.message,
    });
  }

  return {
    schemaVersion: '2',
    thresholds,
    projectRoot: input.cwd.split(path.sep).join('/'),
    framework: {
      name: 'Gnosis',
      version: input.gnosisVersion,
      branding: {
        homepageUrl: 'https://github.com/forkjoin-ai/gnosis',
      },
    },
    config: {
      gnosisMode: input.mode,
      reportKind: 'gnosis-formal-gate',
    },
    files: {
      [key]: {
        language: 'plaintext',
        source: input.source,
        mutants,
      },
    },
  };
}

function parseTypeScriptViolationLine(violation: string): {
  line: number;
  mutatorName: string;
} {
  const fileLines = /^file-lines\s+\S+\s+/u.exec(violation);
  if (fileLines) {
    return { line: 1, mutatorName: 'file-lines' };
  }
  const m = /^(\w+)\s+(.+):(\d+)\s/u.exec(violation);
  if (m?.[1] && m[3]) {
    const line = Number.parseInt(m[3], 10);
    return {
      line: Number.isFinite(line) ? line : 1,
      mutatorName: m[1],
    };
  }
  return { line: 1, mutatorName: 'typescript-threshold' };
}

export interface BuildGnosisTypeScriptMutationReportInput {
  readonly filePath: string;
  readonly source: string;
  readonly cwd: string;
  readonly violations: readonly string[];
  readonly gnosisVersion?: string;
  readonly thresholds?: { high: number; low: number };
}

/**
 * Builds the Gnosis Type Script Mutation Report.
 */
export function buildGnosisTypeScriptMutationReport(
  input: BuildGnosisTypeScriptMutationReportInput
): MutationTestingReport {
  const thresholds = input.thresholds ?? { high: 80, low: 60 };
  const key = reportFileKey(input.filePath, input.cwd);
  const mutants: MutantResult[] = input.violations.map((v, index) => {
    const { line, mutatorName } = parseTypeScriptViolationLine(v);
    return {
      id: `gnosis-ts-${index + 1}`,
      mutatorName,
      location: locationFromLineColumn(line, 1),
      status: 'Killed' as const,
      description: v,
      statusReason: v,
    };
  });

  const ext = path.extname(input.filePath).toLowerCase();
  const language =
    ext === '.tsx' || ext === '.jsx'
      ? 'typescript'
      : ext === '.js'
        ? 'javascript'
        : 'typescript';

  return {
    schemaVersion: '2',
    thresholds,
    projectRoot: input.cwd.split(path.sep).join('/'),
    framework: {
      name: 'Gnosis',
      version: input.gnosisVersion,
      branding: {
        homepageUrl: 'https://github.com/forkjoin-ai/gnosis',
      },
    },
    config: {
      gnosisMode: 'typescript-lint',
      reportKind: 'gnosis-typescript-sonar',
    },
    files: {
      [key]: {
        language,
        source: input.source,
        mutants,
      },
    },
  };
}
