// @a0n/gnosis/src/testing/registry.ts
import { createRequire } from 'node:module';
import type {
  GnosisHookKind,
  GnosisTestCaseDefinition,
  GnosisTestFn,
  GnosisTestSuiteDefinition,
} from './types';
import { unsupportedFeature } from './types';

const ACTIVE_REGISTRY_SYMBOL = Symbol.for('gnosis.testing.activeRegistry');

function resolveRequireBase(): string | null {
  if (typeof __filename === 'string' && __filename.length > 0) {
    return __filename;
  }

  const moduleFilename =
    typeof module !== 'undefined' &&
    module !== null &&
    typeof module.filename === 'string' &&
    module.filename.length > 0
      ? module.filename
      : null;
  if (moduleFilename !== null) {
    return moduleFilename;
  }

  const moduleUrl = import.meta.url;
  return typeof moduleUrl === 'string' && moduleUrl.length > 0
    ? moduleUrl
    : null;
}

const nodeRequireBase = resolveRequireBase();
const nodeRequire =
  nodeRequireBase === null ? null : createRequire(nodeRequireBase);

type ExternalTestBindings = Partial<
  Record<GnosisHookKind | 'describe' | 'it' | 'test', (...args: unknown[]) => unknown>
>;

let cachedExternalBindings: ExternalTestBindings | null | undefined;

function resolveExternalBindings(): ExternalTestBindings | null {
  if (cachedExternalBindings !== undefined) {
    return cachedExternalBindings;
  }

  if (nodeRequire === null) {
    cachedExternalBindings = null;
    return cachedExternalBindings;
  }

  try {
    cachedExternalBindings = nodeRequire('bun:test') as ExternalTestBindings;
    return cachedExternalBindings;
  } catch {
    // continue
  }

  try {
    cachedExternalBindings = nodeRequire('@jest/globals') as ExternalTestBindings;
    return cachedExternalBindings;
  } catch {
    cachedExternalBindings = null;
    return cachedExternalBindings;
  }
}

function createHooks(): Record<GnosisHookKind, GnosisTestFn[]> {
  return {
    beforeAll: [],
    afterAll: [],
    beforeEach: [],
    afterEach: [],
  };
}

function globalRegistryStore(): Record<PropertyKey, unknown> {
  return globalThis as Record<PropertyKey, unknown>;
}

function validateName(kind: string, value: string): void {
  if (typeof value !== 'string' || value.trim().length === 0) {
    throw new Error(`${kind} name must be a non-empty string.`);
  }
}

function validateFn(kind: string, fn: unknown): asserts fn is GnosisTestFn {
  if (typeof fn !== 'function') {
    throw new Error(`${kind} callback must be a function.`);
  }
}

export class GnosisTestRegistry {
  private readonly rootSuite: GnosisTestSuiteDefinition;
  private readonly suiteStack: GnosisTestSuiteDefinition[];
  private declarationCursor = 0;

  public readonly filePath: string;

  public constructor(filePath: string) {
    this.filePath = filePath;
    this.rootSuite = {
      type: 'suite',
      name: '__root__',
      suitePath: [],
      declaredAt: this.nextOrder(),
      items: [],
      hooks: createHooks(),
    };
    this.suiteStack = [this.rootSuite];
  }

  public getRootSuite(): GnosisTestSuiteDefinition {
    return this.rootSuite;
  }

  public getCurrentSuite(): GnosisTestSuiteDefinition {
    return this.suiteStack[this.suiteStack.length - 1]!;
  }

  public addHook(kind: GnosisHookKind, fn: GnosisTestFn): void {
    validateFn(kind, fn);
    this.getCurrentSuite().hooks[kind].push(fn);
  }

  public addCase(name: string, fn: GnosisTestFn, ..._rest: unknown[]): void {
    validateName('test', name);
    validateFn('test', fn);
    const parent = this.getCurrentSuite();
    const definition: GnosisTestCaseDefinition = {
      type: 'case',
      name,
      fn,
      suitePath: parent.suitePath,
      declaredAt: this.nextOrder(),
    };
    parent.items.push(definition);
  }

  public describe(name: string, fn: GnosisTestFn): void {
    validateName('describe', name);
    validateFn('describe', fn);
    const parent = this.getCurrentSuite();
    const suite: GnosisTestSuiteDefinition = {
      type: 'suite',
      name,
      suitePath: [...parent.suitePath, name],
      declaredAt: this.nextOrder(),
      items: [],
      hooks: createHooks(),
    };
    parent.items.push(suite);
    this.suiteStack.push(suite);
    try {
      fn();
    } finally {
      this.suiteStack.pop();
    }
  }

  private nextOrder(): number {
    const current = this.declarationCursor;
    this.declarationCursor += 1;
    return current;
  }
}

/**
 * Handles the gnosis set Active Registry workflow.
 */
export function setActiveRegistry(registry: GnosisTestRegistry | null): void {
  const store = globalRegistryStore();
  if (registry === null) {
    delete store[ACTIVE_REGISTRY_SYMBOL];
    return;
  }
  store[ACTIVE_REGISTRY_SYMBOL] = registry;
}

/**
 * Handles the gnosis get Active Registry workflow.
 */
export function getActiveRegistry(): GnosisTestRegistry {
  const active = globalRegistryStore()[ACTIVE_REGISTRY_SYMBOL];
  if (
    typeof active === 'object' &&
    active !== null &&
    'describe' in active &&
    'addCase' in active &&
    'addHook' in active
  ) {
    return active as GnosisTestRegistry;
  }

  try {
    // intentionally empty
  } catch (e) {
    // describe was not in scope
  }

  // Return a Ghost Registry that resolves globals at runtime to prevent evaluation crashes
  return {
    filePath: 'ghost.test.ts',
    describe: (name: string, fn: any) => {
      const g = globalThis as any;
      if (typeof g.describe === 'function') return g.describe(name, fn);
      const external = resolveExternalBindings();
      if (external && typeof external.describe === 'function') {
        return external.describe(name, fn);
      }
      const t = g.test || g.it;
      if (typeof t === 'function') return t(name, fn);
      if (external) {
        const externalTest = external.test || external.it;
        if (typeof externalTest === 'function') {
          return externalTest(name, fn);
        }
      }
      throw new Error('No active Gnosis test registry and no global describe() found.');
    },
    addCase: (name: string, fn: any, ...rest: unknown[]) => {
      const g = globalThis as any;
      const t = g.test || g.it;
      if (typeof t === 'function') return t(name, fn, ...rest);
      const external = resolveExternalBindings();
      const externalTest = external?.test || external?.it;
      if (typeof externalTest === 'function') {
        return externalTest(name, fn, ...rest);
      }
      throw new Error('No active Gnosis test registry and no global test() found.');
    },
    addHook: (kind: string, fn: any) => {
      const g = globalThis as any;
      if (typeof g[kind] === 'function') return g[kind](fn);
      const external = resolveExternalBindings();
      const externalHook = external?.[kind as GnosisHookKind];
      if (typeof externalHook === 'function') {
        return externalHook(fn);
      }
      // Hooks are often optional
    }
  } as any as GnosisTestRegistry;
}

type DescribeFn = ((name: string, fn: GnosisTestFn) => void) & {
  only: (name: string, fn: GnosisTestFn) => never;
  skip: (name: string, fn: GnosisTestFn) => never;
  each: (
    rows: ReadonlyArray<unknown>
  ) => (nameTemplate: string, fn: (...args: any[]) => any) => void;
};

type TestFnShape = ((
  name: string,
  fn: GnosisTestFn,
  ...rest: unknown[]
) => void) & {
  only: (name: string, fn: GnosisTestFn) => never;
  skip: (name: string, fn: GnosisTestFn) => never;
  todo: (name: string) => never;
  each: (
    rows: ReadonlyArray<unknown>
  ) => (nameTemplate: string, fn: (...args: any[]) => any) => void;
};

function formatEachTitle(template: string, args: readonly unknown[]): string {
  let index = 0;
  return template.replace(/%([sdifjopO%])/g, (match, kind: string) => {
    if (kind === '%') {
      return '%';
    }
    if (index >= args.length) {
      return match;
    }
    const value = args[index];
    index += 1;
    switch (kind) {
      case 's':
        return typeof value === 'string' ? value : String(value);
      case 'd':
      case 'i':
        return String(Math.trunc(Number(value)));
      case 'f':
        return String(Number(value));
      case 'j':
      case 'o':
      case 'O':
        try {
          return JSON.stringify(value);
        } catch {
          return String(value);
        }
      case 'p':
        return String(value);
      default:
        return match;
    }
  });
}

function buildEachFn(
  register: (name: string, fn: GnosisTestFn) => void
): (
  rows: ReadonlyArray<unknown>
) => (nameTemplate: string, fn: (...args: any[]) => any) => void {
  return (rows) => (nameTemplate, fn) => {
    for (const row of rows) {
      const args = Array.isArray(row) ? row : [row];
      const name = formatEachTitle(nameTemplate, args);
      register(name, () => fn(...(args as unknown[])));
    }
  };
}

const unsupportedDescribe = (feature: string) =>
  ((name: string, fn: GnosisTestFn) => {
    void name;
    void fn;
    return unsupportedFeature(feature);
  }) as (name: string, fn: GnosisTestFn) => never;

const unsupportedTest = (feature: string) =>
  ((name: string, fn?: GnosisTestFn) => {
    void name;
    void fn;
    return unsupportedFeature(feature);
  }) as (name: string, fn?: GnosisTestFn) => never;

export const describe = Object.assign(
  (name: string, fn: GnosisTestFn) => getActiveRegistry().describe(name, fn),
  {
    only: unsupportedDescribe('describe.only'),
    skip: unsupportedDescribe('describe.skip'),
    each: buildEachFn((name, fn) => getActiveRegistry().describe(name, fn)),
  }
) as DescribeFn;

export const test = Object.assign(
  (name: string, fn: GnosisTestFn, ...rest: unknown[]) =>
    getActiveRegistry().addCase(name, fn, ...rest),
  {
    only: unsupportedTest('test.only'),
    skip: unsupportedTest('test.skip'),
    todo: unsupportedTest('test.todo') as (name: string) => never,
    each: buildEachFn((name, fn) => getActiveRegistry().addCase(name, fn)),
  }
) as TestFnShape;

export const it = test;

/**
 * Handles the gnosis before All workflow.
 */
export function beforeAll(fn: GnosisTestFn): void {
  getActiveRegistry().addHook('beforeAll', fn);
}

/**
 * Handles the gnosis after All workflow.
 */
export function afterAll(fn: GnosisTestFn): void {
  getActiveRegistry().addHook('afterAll', fn);
}

/**
 * Handles the gnosis before Each workflow.
 */
export function beforeEach(fn: GnosisTestFn): void {
  getActiveRegistry().addHook('beforeEach', fn);
}

/**
 * Handles the gnosis after Each workflow.
 */
export function afterEach(fn: GnosisTestFn): void {
  getActiveRegistry().addHook('afterEach', fn);
}
