// @a0n/gnosis/src/testing/vitest.ts
import {
  clearAllMocks,
  mock,
  restoreAllMocks,
  spyOn,
  unsupportedFeature,
} from './index';

export {
  afterAll,
  afterEach,
  beforeAll,
  beforeEach,
  describe,
  expect,
  it,
  mock,
  spyOn,
  test,
} from './index';

// Tracks globals replaced via vi.stubGlobal so vi.unstubAllGlobals can restore
// them. Without this, vi.stubGlobal('fetch', mock) silently no-ops and tests
// that mock fetch/globals leak through to the real implementation (e.g. live
// network), producing confusing pass/fail depending on real-world data.
const stubbedGlobals = new Map<string, { existed: boolean; value: unknown }>();

function stubGlobalImpl(name: string, value: unknown): typeof vi {
  const g = globalThis as Record<string, unknown>;
  if (!stubbedGlobals.has(name)) {
    stubbedGlobals.set(name, { existed: name in g, value: g[name] });
  }
  g[name] = value;
  return vi;
}

function unstubAllGlobalsImpl(): typeof vi {
  const g = globalThis as Record<string, unknown>;
  for (const [name, prev] of stubbedGlobals) {
    if (prev.existed) g[name] = prev.value;
    else delete g[name];
  }
  stubbedGlobals.clear();
  return vi;
}

export const vi = Object.freeze({
  fn: mock,
  spyOn,
  clearAllMocks,
  restoreAllMocks,
  stubGlobal: stubGlobalImpl,
  unstubAllGlobals: unstubAllGlobalsImpl,
  mocked<T>(value: T): T {
    return value;
  },
  useFakeTimers(): never {
    return unsupportedFeature('vi.useFakeTimers');
  },
  useRealTimers(): never {
    return unsupportedFeature('vi.useRealTimers');
  },
  advanceTimersByTime(): never {
    return unsupportedFeature('vi.advanceTimersByTime');
  },
  runAllTimers(): never {
    return unsupportedFeature('vi.runAllTimers');
  },
  mock(): never {
    return unsupportedFeature('vi.mock');
  },
  module(
    specifier: string,
    factory: () => unknown | Promise<unknown>,
    importerPath?: string
  ): void {
    return mock.module(specifier, factory, importerPath);
  },
});
