// @a0n/gnosis/third-eye/host/third-eye-host.ts
// third-eye-host.ts — the ThirdEyeHost shim: the isomorphic (Node + Worker) host
// layer that binds the third-eye perception WASM into a service, ZERO-COPY both
// directions. Modeled on distributed-inference-host's HostShim (a thin TS layer
// over the wasm-bindgen surface), but SIMPLER per WASM_PORT_PLAN.md §3: perception
// carries no external weights, so there is no R2/knot/Range-fetch/LRU — the durable
// state (EMA background + tracker + SceneModel + frf value-door + antiqueue
// corridor) lives INSIDE the wasm `PerceptionEngine` instance. One create() + one
// hot observe() + dispose().
//
// THE ZERO-COPY CONTRACT (mirrors wasm_bindings.rs §"Zero-copy contract"):
//   - the engine OWNS the RGB24 frame buffer and the packed-tick output buffer in
//     wasm linear memory;
//   - the host writes each frame DIRECTLY into the engine's frame buffer via a
//     cached Uint8Array VIEW over [framePtr, framePtr+frameCapacity) — `view.set(rgb)`
//     is the ONLY frame-sized write per frame; no `&[u8]` frame argument crosses the
//     bindgen boundary (observe takes one f64 `ptsS`);
//   - the host reads the packed `ConsciousTick` wire bytes back via a VIEW over
//     [tickPtr, tickPtr+tickLen) — zero copy OUT; the caller forwards that view
//     straight onto the WS.
//
// RETAIN SEMANTICS (documented, load-bearing): `observe()` returns a Uint8Array
// VIEW into wasm linear memory. It is valid ONLY until the next `observe()` (which
// overwrites the tick buffer) or any wasm allocation that grows linear memory
// (which detaches `memory.buffer`). A caller that must retain the tick past the next
// observe MUST copy it (`new Uint8Array(view)` or pass { copy: true }). The hot WS
// path forwards-and-forgets within the same tick, so it takes the zero-copy view.
//
// ISOMORPHIC LOADING:
//   - Node: pass nothing — the shim reads pkg/third_eye_bg.wasm off disk + the
//     pkg's `initSync` (synchronous, no fetch). Or call ThirdEyeHost.load() once to
//     pre-init and share the module across engines.
//   - Worker/browser: pass an already-initialized wasm module (the `InitOutput` the
//     pkg's default async init returns, which exposes `.memory`), OR call
//     ThirdEyeHost.loadFromInit(initFn) with the pkg's default export. The wasm is
//     passed IN (di-host doctrine: "callers pass the wasm-bindgen module in
//     directly"), so the shim runs on CF Workers, Node, Bun, or Deno.
//
// NO ML, no paid services, no fabrication. The shim moves bytes and sequences calls;
// all perception is the sovereign third-eye wasm.

// The pkg's TS surface. Resolved relative to this file (../pkg) so the shim has no
// build-step dependency — Node 22+/25 strips the .ts on import, and the pkg .js is
// plain ESM. `PerceptionEngine` + the init functions are the only symbols used.
import {
  PerceptionEngine,
  __te_init,
  initSync,
  default as initAsync,
} from '../pkg/third_eye.js';
import type { InitOutput } from '../pkg/third_eye.js';

// The minimal wasm-module surface the shim needs to do zero-copy: the linear memory
// (for the frame/tick views) and the `__wbindgen_start` already ran via init. We
// only ever read `memory.buffer`. `PerceptionEngine` itself comes from the pkg
// import (a wasm-bindgen class bound to the same module instance after init).
export interface ThirdEyeWasm {
  readonly memory: WebAssembly.Memory;
}

export interface ThirdEyeOpts {
  /** a fixed desk/laptop webcam is static; an orbiting studio render is not. */
  readonly staticCam?: boolean;
  /** object priors as a JSON string — the host read the file (lifts priors.rs FS off wasm). */
  readonly objectPriorsJson?: string;
  readonly cMax?: number;
  readonly handReach?: number;
  readonly maxImpulse?: number;
  readonly plateauWindow?: number;
  readonly worldW?: number;
  readonly worldD?: number;
  /** frf value-door per-item work estimate (ns); host-tunable, telemetry-only on wasm. */
  readonly workNs?: number;
  readonly dispatchNs?: number;
  readonly sourceId?: string;
  readonly layerId?: string;
}

/** antiqueue corridor + frf value-door telemetry from the most recent observe. */
export interface ThirdEyeTelemetry {
  /** gnosis-frf value-door verdict for the per-frame stat work ("GREEN"/"RED"). */
  readonly governorLight: string;
  /** antiqueue corridor: frame-events published this session (throughput signal). */
  readonly published: number;
  /** antiqueue corridor: peak in-flight (back-pressure signal — the host can throttle). */
  readonly peakInFlight: number;
  /** frame-to-frame motion energy in [0,1] (the governor's signal). */
  readonly motion: number;
  /** fully-certified (Accept) entities in the current scene. */
  readonly accepted: number;
  /** the next sequence number that will be stamped onto a tick. */
  readonly sequence: number;
}

/** Module-level singleton wasm instance for the Node path: init the pkg ONCE and
 *  share `memory` across engines. Worker callers inject their own (see fromWasm). */
let nodeWasm: ThirdEyeWasm | null = null;

/** Node-only synchronous wasm init: read pkg/third_eye_bg.wasm off disk and
 *  `initSync`. Idempotent (the pkg's initSync no-ops if already initialized).
 *  Throws a clear error in a non-Node runtime (use fromWasm/loadFromInit there). */
async function ensureNodeWasm(): Promise<ThirdEyeWasm> {
  if (nodeWasm) return nodeWasm;
  // Dynamic imports so this module loads in a Worker bundle that has no node:fs.
  const { readFileSync } = await import('node:fs');
  const { fileURLToPath } = await import('node:url');
  const wasmUrl = new URL('../pkg/third_eye_bg.wasm', import.meta.url);
  const bytes = readFileSync(fileURLToPath(wasmUrl));
  const out = initSync({ module: bytes }) as unknown as ThirdEyeWasm;
  nodeWasm = out;
  return out;
}

/**
 * ThirdEyeHost — one perception SESSION bound to one wasm `PerceptionEngine`. Create
 * one per stream (per WS connection), feed it frames hot, dispose at end-of-stream.
 */
export class ThirdEyeHost {
  private readonly engine: PerceptionEngine;
  private readonly wasm: ThirdEyeWasm;
  private readonly frameCap: number;
  private readonly framePtr: number;
  /** Cached view over the engine's frame buffer in wasm linear memory. Re-derived
   *  only when `memory.buffer` detaches (a wasm growth elsewhere moves the backing). */
  private frameView: Uint8Array;
  private disposed = false;

  private constructor(engine: PerceptionEngine, wasm: ThirdEyeWasm) {
    this.engine = engine;
    this.wasm = wasm;
    this.frameCap = engine.frameCapacity();
    this.framePtr = engine.framePtr();
    this.frameView = new Uint8Array(wasm.memory.buffer, this.framePtr, this.frameCap);
  }

  /** Node path: init the pkg wasm off disk (once, shared) and create an engine. */
  static async create(
    width: number,
    height: number,
    fps: number,
    opts: ThirdEyeOpts = {},
  ): Promise<ThirdEyeHost> {
    const wasm = await ensureNodeWasm();
    return ThirdEyeHost.fromWasm(width, height, fps, opts, wasm);
  }

  /** Worker/browser path: the caller already initialized the pkg wasm (its default
   *  async init returns `InitOutput`, which exposes `.memory`) and passes it in. The
   *  pkg's `PerceptionEngine` class is bound to that same instance after init. */
  static fromWasm(
    width: number,
    height: number,
    fps: number,
    opts: ThirdEyeOpts,
    wasm: ThirdEyeWasm | InitOutput,
  ): ThirdEyeHost {
    __te_init(); // idempotent panic-hook install
    const engine = new PerceptionEngine(width, height, fps, JSON.stringify(opts ?? {}));
    return new ThirdEyeHost(engine, wasm as ThirdEyeWasm);
  }

  /** Worker/browser convenience: run the pkg's default async init (fetch/instantiate)
   *  and cache the module so subsequent fromWasm/observe share it. `input` is whatever
   *  the pkg's default init accepts (a URL/Response/bytes/Module); omit on bundlers
   *  that inline the wasm. */
  static async loadFromInit(input?: unknown): Promise<ThirdEyeWasm> {
    const out = (await initAsync(input as never)) as unknown as ThirdEyeWasm;
    return out;
  }

  /**
   * Perceive ONE frame, zero-copy both directions.
   *
   * 1. write `rgb24` (length === frameCapacity) DIRECTLY into the engine's frame
   *    buffer via the cached view — the only frame-sized write;
   * 2. `observe(ptsS)` runs the in-place SIMD / frf value-door / antiqueue-corridor
   *    pipeline with no frame argument and no per-frame frame-sized allocation;
   * 3. return a VIEW over the packed `ConsciousTick` wire bytes (10-byte FlowCodec
   *    header + JSON payload) — the EXACT bytes the WS forwards.
   *
   * The returned view aliases wasm linear memory and is valid only until the next
   * `observe()`. Pass `{ copy: true }` to get an owned copy when the caller must
   * retain it (e.g. buffering ticks). The hot WS path forwards within the tick and
   * takes the view (no copy).
   */
  observe(rgb24: Uint8Array, ptsS: number, options?: { copy?: boolean }): Uint8Array {
    if (this.disposed) throw new Error('ThirdEyeHost.observe after dispose');
    if (rgb24.length !== this.frameCap) {
      throw new RangeError(
        `ThirdEyeHost.observe: frame is ${rgb24.length}B, expected ${this.frameCap}B ` +
          `(width*height*3). Strip alpha (RGBA -> RGB24) and remove row padding.`,
      );
    }
    // Re-derive the frame view if a prior allocation grew linear memory (buffer moved).
    if (this.frameView.buffer !== this.wasm.memory.buffer || this.frameView.byteLength === 0) {
      this.frameView = new Uint8Array(this.wasm.memory.buffer, this.framePtr, this.frameCap);
    }
    this.frameView.set(rgb24); // ZERO-COPY IN — straight into wasm memory
    const len = this.engine.observe(ptsS); // perceive in place; no frame arg
    const tickPtr = this.engine.tickPtr();
    // ZERO-COPY OUT — a view over the packed tick. observe() can grow linear memory
    // (the tick Vec may reallocate), so read memory.buffer fresh here.
    const view = new Uint8Array(this.wasm.memory.buffer, tickPtr, len);
    return options?.copy ? new Uint8Array(view) : view;
  }

  /** DEBUG accessor (NOT the hot path): the human-readable relay JSON for the
   *  last-perceived scene state. Allocates a String across the boundary — avoid per
   *  frame. Use the binary observe() return for the wire. */
  tickJson(ptsS: number): string {
    if (this.disposed) throw new Error('ThirdEyeHost.tickJson after dispose');
    return this.engine.tickJson(ptsS);
  }

  /** antiqueue + frf telemetry from the most recent observe. */
  telemetry(): ThirdEyeTelemetry {
    if (this.disposed) throw new Error('ThirdEyeHost.telemetry after dispose');
    return {
      governorLight: this.engine.governorLight(),
      published: this.engine.published(),
      peakInFlight: this.engine.peakInFlight(),
      motion: this.engine.motion(),
      accepted: this.engine.accepted(),
      sequence: this.engine.sequence(),
    };
  }

  /** Free the wasm engine (its frame/tick buffers + PerceptionCore). Idempotent. */
  dispose(): void {
    if (this.disposed) return;
    this.disposed = true;
    try {
      this.engine.dispose();
    } catch {
      // engine may already be moved-out; free() is the fallback the bindgen generated.
    }
  }
}
