// @a0n/.ignored_gnosis/distributed-inference-host/__tests__/fixtures.ts
/**
 * Test fixtures: a tiny in-memory knot, a memory-backed KnotByteSource that
 * counts calls, and a mock WasmModule that records every backend mutation.
 *
 * The fixtures don't run real inference — they just verify the shim's
 * orchestration: header parse, fetch fan-out, set_tensor / evict ordering.
 */

import type {
  KnotByteSource,
  KnotMetadata,
  TensorInfo,
  WasmHostBackendLike,
  WasmLlamaPipelineLike,
  WasmModule,
} from '../src/types';
import { computePayloadOffset } from '../src/knot-header';
import { tensorNamesForLayer, RESIDENT_TENSOR_NAMES } from '../src/tensor-names';

const KNOT_MAGIC_BYTES = new TextEncoder().encode('KNOT');

export interface BuildKnotOptions {
  numLayers: number;
  /** Bytes per per-layer tensor. Tiny by default for fast tests. */
  layerTensorBytes?: number;
  /** Bytes per resident tensor. */
  residentTensorBytes?: number;
  /** Whether to include the optional output_weight tensor. */
  includeOutputWeight?: boolean;
  /** Knot file format version. */
  version?: number;
  /** Codec marker byte. */
  codecMarker?: number;
}

export interface BuiltKnot {
  bytes: Uint8Array;
  metadata: KnotMetadata;
  payloadOffset: number;
  layerTensorBytes: number;
  residentTensorBytes: number;
}

/**
 * Build a synthetic knot in memory. The tensor payload is filled with
 * deterministic content per tensor (the byte value cycles with the tensor
 * name's hash) so tests can assert which bytes flowed where.
 */
export function buildFakeKnot(opts: BuildKnotOptions): BuiltKnot {
  const numLayers = opts.numLayers;
  const layerTensorBytes = opts.layerTensorBytes ?? 16;
  const residentTensorBytes = opts.residentTensorBytes ?? 8;
  const includeOutputWeight = opts.includeOutputWeight ?? true;
  const version = opts.version ?? 1;
  const codecMarker = opts.codecMarker ?? 0;

  // Build the tensor table with running offsets relative to payload start.
  const tensors: Record<string, TensorInfo> = {};
  let cursor = 0;
  const pushTensor = (name: string, length: number, type: string): void => {
    tensors[name] = { offset: cursor, length, tensor_type: type };
    cursor += length;
  };

  // Residents come first.
  pushTensor('token_embd_weight', residentTensorBytes, 'F16');
  pushTensor('output_norm_weight', residentTensorBytes, 'F32');
  if (includeOutputWeight) {
    pushTensor('output_weight', residentTensorBytes, 'F16');
  }

  // Per-layer tensors.
  for (let l = 0; l < numLayers; l += 1) {
    for (const name of tensorNamesForLayer(l)) {
      pushTensor(name, layerTensorBytes, 'Q4_K');
    }
  }

  const totalPayloadBytes = cursor;

  const metadata: KnotMetadata = {
    config: { num_layers: numLayers },
    tensors,
  };
  const metadataJson = JSON.stringify(metadata);
  const metadataBytes = new TextEncoder().encode(metadataJson);
  const metadataLength = metadataBytes.length;

  const payloadOffset = computePayloadOffset(metadataLength);
  const totalSize = payloadOffset + totalPayloadBytes;

  const out = new Uint8Array(totalSize);
  // Magic
  out.set(KNOT_MAGIC_BYTES, 0);
  // Version
  const headerView = new DataView(out.buffer, out.byteOffset, out.byteLength);
  headerView.setUint16(4, version, true);
  // Metadata length
  headerView.setUint32(6, metadataLength, true);
  // Metadata JSON
  out.set(metadataBytes, 10);
  // (padding bytes are already 0 from Uint8Array zero-init)
  // Codec marker
  out[payloadOffset - 1] = codecMarker;

  // Fill tensor payload deterministically.
  for (const [name, info] of Object.entries(tensors)) {
    const seed = simpleHash(name);
    const start = payloadOffset + info.offset;
    for (let i = 0; i < info.length; i += 1) {
      out[start + i] = (seed + i) & 0xff;
    }
  }

  return {
    bytes: out,
    metadata,
    payloadOffset,
    layerTensorBytes,
    residentTensorBytes,
  };
}

function simpleHash(s: string): number {
  let h = 2166136261 >>> 0;
  for (let i = 0; i < s.length; i += 1) {
    h ^= s.charCodeAt(i);
    h = Math.imul(h, 16777619) >>> 0;
  }
  return h & 0xff;
}

/**
 * Memory-backed KnotByteSource that records every range request so tests
 * can assert call counts and order.
 */
export class RecordingKnotSource implements KnotByteSource {
  readonly calls: { start: number; end: number; length: number }[] = [];

  constructor(private readonly buf: Uint8Array) {}

  async rangeGet(start: number, end: number): Promise<Uint8Array> {
    if (start < 0 || end < start) {
      throw new Error(`RecordingKnotSource: bad range ${start}..${end}`);
    }
    const clampedEnd = Math.min(end, this.buf.length - 1);
    const slice = this.buf.subarray(start, clampedEnd + 1);
    // Copy so the caller can't accidentally mutate fixture bytes.
    const out = new Uint8Array(slice);
    this.calls.push({ start, end, length: out.length });
    return out;
  }

  totalBytesRead(): number {
    return this.calls.reduce((sum, c) => sum + c.length, 0);
  }

  reset(): void {
    this.calls.length = 0;
  }
}

/** Recorded set_tensor call. */
export interface SetTensorCall {
  name: string;
  bytes: Uint8Array;
}

/** Recorded evict call. */
export interface EvictCall {
  names: string[];
}

/**
 * Mock WasmHostBackend. Implements the structural interface and records every
 * mutation. tensor_info answers from the metadata embedded at construction.
 */
export class MockBackend implements WasmHostBackendLike {
  readonly setCalls: SetTensorCall[] = [];
  readonly evictCalls: EvictCall[] = [];
  readonly tensorTable: Record<string, TensorInfo>;
  freed = false;

  constructor(metadataJson: string) {
    const parsed = JSON.parse(metadataJson) as
      | KnotMetadata
      | Record<string, TensorInfo>;
    // The shim might pass either the full metadata or just the tensors map.
    if ((parsed as KnotMetadata).tensors) {
      this.tensorTable = (parsed as KnotMetadata).tensors;
    } else {
      this.tensorTable = parsed as Record<string, TensorInfo>;
    }
  }

  set_tensor(name: string, bytes: Uint8Array): void {
    // Store a copy so subsequent buffer reuse can't pollute the recording.
    this.setCalls.push({ name, bytes: new Uint8Array(bytes) });
  }

  evict(names: string[]): void {
    this.evictCalls.push({ names: names.slice() });
  }

  tensor_info(name: string): TensorInfo | null {
    return this.tensorTable[name] ?? null;
  }

  free(): void {
    this.freed = true;
  }
}

export class MockPipeline implements WasmLlamaPipelineLike {
  pos = 0;
  readonly generateCalls: { prompt: Uint32Array; max: number }[] = [];
  freed = false;

  generate(prompt_tokens: Uint32Array, max_new_tokens: number): Uint32Array {
    this.generateCalls.push({ prompt: new Uint32Array(prompt_tokens), max: max_new_tokens });
    this.pos = prompt_tokens.length + max_new_tokens;
    // Return some canned tokens so callers can assert it propagates.
    const out = new Uint32Array(max_new_tokens);
    for (let i = 0; i < max_new_tokens; i += 1) out[i] = 1000 + i;
    return out;
  }

  get_pos(): number {
    return this.pos;
  }

  free(): void {
    this.freed = true;
  }
}

/**
 * Build a complete mock WasmModule. The pipeline produced by from_backend is
 * exposed via `.pipeline` on the module for assertions, since wasm-bindgen
 * does not give callers a handle to "the most recent pipeline" otherwise.
 */
export function buildMockWasmModule(): WasmModule & {
  lastBackend: MockBackend | null;
  lastPipeline: MockPipeline | null;
} {
  const mod = {
    lastBackend: null as MockBackend | null,
    lastPipeline: null as MockPipeline | null,
    WasmHostBackend: class extends MockBackend {
      constructor(metadataJson: string) {
        super(metadataJson);
        mod.lastBackend = this;
      }
    },
    WasmLlamaPipeline: {
      from_backend(
        _backend: WasmHostBackendLike,
        _configJson: string | null,
        _payloadOffset: number,
        _draftMode: boolean,
        _resonance: number,
      ): WasmLlamaPipelineLike {
        const p = new MockPipeline();
        mod.lastPipeline = p;
        return p;
      },
    },
  };
  return mod as typeof mod & WasmModule;
}

/** Convenience: 9 names per layer, mirrors the production helper. */
export function expectedLayerNames(layerIdx: number): string[] {
  return tensorNamesForLayer(layerIdx);
}

/** Convenience: residents that should always show up regardless of layers. */
export function expectedResidentNames(includeOutputWeight: boolean): string[] {
  if (includeOutputWeight) return [...RESIDENT_TENSOR_NAMES];
  return RESIDENT_TENSOR_NAMES.filter((n) => n !== 'output_weight');
}
