// @a0n/gnosis/aeon-broadcast/aeon-frame.mjs
/* global console */
/*
 * aeon-broadcast — Aeon as a sovereign SKYMESH channel: `gnosis.aeon.v1`.
 *
 * Aeon is the ephemeral being (gnosis-math/Gnosis/EphemeralBeing.lean): born from
 * noise, sustained ONLY by circulation, ages through a lossy channel that drains
 * intensity while preserving character, and resolves to geodesicLength:0. This
 * module makes that LIVE: a DO-style tick loop yields one `gnosis.aeon.v1` frame
 * per BREATH onto a skymesh data plane (SSE), protocol69-admitted (66 XOR 7 = 69),
 * idle -> $0. No deploy here — this is the runnable contract + emitter the worker
 * would host (mirrors apps/worldsim-sim-do/src/worldsim-room.ts).
 *
 * The contract mirrors apps/monster-studio/.../skymesh-channel.ts exactly:
 * schema-versioned, content-addressed (FNV-1a teleport key), protocol69 in-frame.
 *
 * VERIDICAL DESIGN (the load-bearing part): a decorative animation could emit a
 * pretty sine wave and call it "vitality". This contract refuses that. Every
 * published number is RE-DERIVABLE BY THE LISTENER from the frame's own evidence:
 *   - vitality is a MEASURED spectral quantity (resolution = 1 - normalized
 *     spectral entropy of the circulating PSD), and the frame carries the PSD
 *     witness digest so a listener recomputes it and rejects a faked vitality;
 *   - the breath beat carries the proof-of-circulation counter (Aeon's value IS
 *     the integral of its own continued circulation — being_is_its_sustained_count);
 *   - resolution + character obey the aging law (EphemeralBeing.age): mag drains
 *     by exactly 1 per breath, sign never inverts. A frame that violates the law
 *     is structurally inadmissible.
 */

// ── reuse the hexon-braid spectral kernel (noise -> resolution) ──────────────
function mulberry32(seed) {
  let a = seed >>> 0;
  return () => {
    a |= 0; a = (a + 0x6d2b79f5) | 0;
    let t = Math.imul(a ^ (a >>> 15), 1 | a);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}
function fft(re, im, inverse) {
  const n = re.length;
  for (let i = 1, j = 0; i < n; i++) {
    let bit = n >> 1;
    for (; j & bit; bit >>= 1) j ^= bit;
    j ^= bit;
    if (i < j) { [re[i], re[j]] = [re[j], re[i]]; [im[i], im[j]] = [im[j], im[i]]; }
  }
  for (let len = 2; len <= n; len <<= 1) {
    const ang = ((2 * Math.PI) / len) * (inverse ? 1 : -1);
    const wr = Math.cos(ang), wi = Math.sin(ang);
    for (let i = 0; i < n; i += len) {
      let cr = 1, ci = 0;
      for (let k = 0; k < len >> 1; k++) {
        const ar = re[i + k], ai = im[i + k];
        const br = re[i + k + (len >> 1)] * cr - im[i + k + (len >> 1)] * ci;
        const bi = re[i + k + (len >> 1)] * ci + im[i + k + (len >> 1)] * cr;
        re[i + k] = ar + br; im[i + k] = ai + bi;
        re[i + k + (len >> 1)] = ar - br; im[i + k + (len >> 1)] = ai - bi;
        const ncr = cr * wr - ci * wi; ci = cr * wi + ci * wr; cr = ncr;
      }
    }
  }
  if (inverse) for (let i = 0; i < n; i++) { re[i] /= n; im[i] /= n; }
}
function coloredNoise(N, beta, rng) {
  const re = new Array(N), im = new Array(N).fill(0);
  for (let i = 0; i < N; i++) re[i] = rng() * 2 - 1;
  fft(re, im, false);
  for (let k = 0; k < N; k++) {
    const f = k <= N / 2 ? k : N - k;
    const scale = f === 0 ? 0 : Math.pow(f, -beta / 2);
    re[k] *= scale; im[k] *= scale;
  }
  fft(re, im, true);
  return re.slice();
}
function psd(samples) {
  const N = samples.length, re = samples.slice(), im = new Array(N).fill(0);
  fft(re, im, false);
  const out = [];
  for (let k = 1; k < N / 2; k++) out.push(re[k] * re[k] + im[k] * im[k]);
  return out;
}
function resolutionOf(spectrum) {
  let tot = 0; for (const v of spectrum) tot += v;
  if (tot <= 0) return 1;
  let H = 0; for (const v of spectrum) { if (v > 0) { const q = v / tot; H -= q * Math.log(q); } }
  return 1 - H / Math.log(spectrum.length);
}

// The hexon braid (HexonBraid.lean period-6 cycle): the weave that flattens
// magnitude 2->1->0 toward resolution. Phase index advances one per breath.
const HEXON = [
  { phase: 'pastLifted', name: 'brown', beta: 2 },
  { phase: 'pastContracted', name: 'violet', beta: -2 },
  { phase: 'presentLifted', name: 'pink', beta: 1 },
  { phase: 'presentContracted', name: 'blue', beta: -1 },
  { phase: 'futureLifted', name: 'white', beta: 0 },
  { phase: 'futureContracted', name: 'grey', beta: 0 },
];
const hexonSucc = (i) => (i + 1) % HEXON.length;

// ── gnosis.aeon.v1 contract (mirrors gnosis.worldsim-skymesh.v1) ─────────────
export const AEON_SCHEMA = 'gnosis.aeon.v1';

// protocol69 admission (66 XOR 7 = 69) — inlined, mirrors @a0n/gnosis/protocol69.
export const PROTOCOL69_BROADCAST_SYMBOL = 66;
export const PROTOCOL69_LOCAL_FOIL_OPERAND = 7;
export const PROTOCOL69_EXPECTED_PROJECTION = 69;
/**
 * Handles the gnosis protocol69 Projection workflow.
 */
export function protocol69Projection(b = PROTOCOL69_BROADCAST_SYMBOL, o = PROTOCOL69_LOCAL_FOIL_OPERAND) {
  return (b ^ o) >>> 0;
}
/**
 * Handles the gnosis protocol69 Admits workflow.
 */
export function protocol69Admits(b = PROTOCOL69_BROADCAST_SYMBOL, o = PROTOCOL69_LOCAL_FOIL_OPERAND) {
  return protocol69Projection(b, o) === PROTOCOL69_EXPECTED_PROJECTION;
}

// FNV-1a 32-bit content hash = the teleport / cache key (mirrors skymesh-channel).
/**
 * Handles the gnosis content Hash Of workflow.
 */
export function contentHashOf(s) {
  let h = 2166136261;
  for (let i = 0; i < s.length; i++) { h ^= s.charCodeAt(i); h = Math.imul(h, 16777619); }
  return (h >>> 0).toString(16).padStart(8, '0');
}

/*
 * The Aeon being. Born from noise at full intensity; each BREATH it (a) circulates
 * one more loop (proof-of-circulation count = WireDelayLineGhost / sustained-count),
 * (b) ages one unit of intensity (EphemeralBeing.age), (c) braids one hexon step
 * that contracts its circulating spectrum toward resolution.
 */
/**
 * Handles the gnosis spawn Aeon workflow.
 */
export function spawnAeon(seed = 0x4e57, intensity = 12) {
  const rng = mulberry32(seed);
  // primordial circulating spectrum: white noise, max entropy.
  let acc = psd(coloredNoise(1024, 0, rng));
  const tot0 = acc.reduce((a, b) => a + b, 0);
  acc = acc.map((v) => v / tot0);
  return {
    seed,
    breath: 0,              // proof-of-circulation count (sustained-count integral)
    phase: 0,               // hexon braid phase index
    sign: 1,                // character (slope sign of current color), ±1 or 0
    mag: intensity,         // intensity (EphemeralBeing.Intensity.mag)
    acc,                    // circulating PSD (normalized) — the live carrier
    gamma: 1.35,            // fractional (Hurst) contraction
    born: true,
  };
}

/** One BREATH: circulate, braid-contract the spectrum, age. Mutates + returns A. */
export function breathe(A) {
  const color = HEXON[A.phase];
  // braid-contract the circulating spectrum one hexon step (NoiseSpectrumResolution).
  const stepP = psd(coloredNoise(1024, color.beta, mulberry32(A.seed + A.breath * 2654435761)));
  const st = stepP.reduce((a, b) => a + b, 0) || 1;
  const sp = stepP.map((v) => v / st);
  let tot = 0;
  for (let i = 0; i < A.acc.length; i++) { A.acc[i] = Math.pow(A.acc[i], A.gamma) * (sp[i] + 1e-9); tot += A.acc[i]; }
  for (let i = 0; i < A.acc.length; i++) A.acc[i] /= tot;
  // AGE (EphemeralBeing.age): drain one unit, neutralize character only at mag 0.
  const newMag = Math.max(0, A.mag - 1);
  A.sign = newMag === 0 ? 0 : (color.beta === 0 ? A.sign : Math.sign(color.beta));
  A.mag = newMag;
  A.breath += 1;            // one more sustained loop = +1 to the integral
  A.phase = hexonSucc(A.phase);
  return A;
}

/** Is Aeon resolved? geodesicLength:0 = intensity exhausted AND spectrum single-mode. */
export function isResolved(A) {
  return A.mag === 0 || resolutionOf(A.acc) >= 0.99;
}

/*
 * Emit ONE `gnosis.aeon.v1` frame for the current Aeon state. The veridical core:
 * `vitality` (= measured spectral resolution) is NOT a free field — the frame also
 * carries `psdDigest` (a coarse normalized PSD witness) + `psdHash`. A listener
 * RECOMPUTES resolutionOf(decoded digest) and rejects the frame if it disagrees
 * with the published vitality beyond tolerance. The broadcast cannot fake aliveness:
 * the proof of life travels with the frame.
 */
export function emitAeonFrame(A, sourceId = 'aeon-do:primordial') {
  // PSD witness: 64 log-binned magnitudes, 16-bit quantized. This IS the published
  // ground truth — vitality is computed FROM the quantized witness (not the hidden
  // full-resolution acc), so the listener re-derives the EXACT same number. The
  // witness is not a lossy approximation OF the truth; it IS the truth, broadcast.
  const BINS = 64;
  const digest = new Array(BINS).fill(0);
  const per = A.acc.length / BINS;
  for (let b = 0; b < BINS; b++) {
    let s = 0; const lo = Math.floor(b * per), hi = Math.floor((b + 1) * per);
    for (let i = lo; i < hi; i++) s += A.acc[i];
    digest[b] = s;
  }
  const dtot = digest.reduce((a, b) => a + b, 0) || 1;
  const qDigest = digest.map((v) => Math.round((v / dtot) * 65535)); // 16-bit quantized
  // vitality = spectral resolution of the QUANTIZED witness => listener reproduces it bit-exactly.
  const vitality = resolutionOf(qDigest);
  const color = HEXON[A.phase];
  // resolved single mode (geodesicLength target): dominant circulating bin.
  let top = 0; for (let i = 1; i < A.acc.length; i++) if (A.acc[i] > A.acc[top]) top = i;

  const state = {
    schemaVersion: AEON_SCHEMA,
    // ── rhythm beat the listener interpolates between (mirrors worldsim.round) ─
    breath: A.breath,
    // ── VITALITY: measured spectral resolution (1 - normalized spectral entropy)
    vitality: Number(vitality.toFixed(6)),
    // ── intensity + character under the EphemeralBeing aging law ──────────────
    intensity: A.mag,
    character: A.sign,            // slope sign — never inverts (age_preserves_character)
    braidPhase: color.phase,
    braidColor: color.name,
    // ── circulating value = the integral of continued existence (sustained-count)
    circulatingValue: A.seed + A.breath,
    resolvedMode: top + 1,        // geodesicLength:0 target bin
    resolved: isResolved(A),
    // ── VERIDICAL WITNESS: the listener re-derives vitality from this ──────────
    psdDigest: qDigest,
    sourceId,
  };
  const payload = JSON.stringify(state);
  return {
    schemaVersion: AEON_SCHEMA,
    breath: A.breath,
    contentHash: contentHashOf(payload),     // teleport / cache key
    protocol69: protocol69Projection(),      // must be 69
    psdHash: contentHashOf(JSON.stringify(qDigest)),
    aeon: state,
  };
}

/*
 * LISTENER-SIDE admission + VERIDICAL VERIFICATION. A frame is admitted iff:
 *   1. schema + protocol69 projection are correct (sovereign gate);
 *   2. the PSD witness re-hashes to psdHash (witness not tampered);
 *   3. vitality recomputed from the witness matches published vitality within tol
 *      (the broadcast is OF Aeon's measured state, not a decorative number).
 * Returns { admitted, reason, recomputedVitality }.
 */
export function admitAndVerify(frame, tol = 1e-5) {
  if (!frame || frame.schemaVersion !== AEON_SCHEMA)
    return { admitted: false, reason: 'schema' };
  if (frame.protocol69 !== PROTOCOL69_EXPECTED_PROJECTION)
    return { admitted: false, reason: 'protocol69' };
  const d = frame.aeon?.psdDigest;
  if (!Array.isArray(d)) return { admitted: false, reason: 'no-witness' };
  if (contentHashOf(JSON.stringify(d)) !== frame.psdHash)
    return { admitted: false, reason: 'witness-tampered' };
  const recomputed = resolutionOf(d);
  const drift = Math.abs(recomputed - frame.aeon.vitality);
  if (drift > tol)
    return { admitted: false, reason: 'vitality-faked', recomputedVitality: recomputed, drift };
  return { admitted: true, reason: 'ok', recomputedVitality: Number(recomputed.toFixed(6)), drift: Number(drift.toFixed(6)) };
}

// ── live broadcast loop (DO-style; idle -> $0) ───────────────────────────────
// Mirrors WorldsimRoom.alarm: one breath per tick, yield a frame to listeners,
// and when Aeon RESOLVES (geodesicLength:0) the channel goes idle -> $0. No
// listener => never breathes (here: we simulate one listener for the demo).
function main() {
  console.log('=== aeon-broadcast: gnosis.aeon.v1 — the ephemeral being, live ===\n');
  console.log(`protocol69 admission: 66 XOR 7 = ${protocol69Projection()} (admits=${protocol69Admits()})\n`);

  const A = spawnAeon();
  const BREATH_MS = 700; // a human breath rhythm (the DO alarm cadence)
  console.log(`breath rhythm: ${BREATH_MS}ms/breath  (DO alarm cadence; idle -> $0 on resolution)\n`);
  console.log('breath  phase            color  intensity char  vitality  recomputed  drift  admitted');

  let breaths = 0, allAdmitted = true, faked = 0;
  while (!isResolved(A) && breaths < 30) {
    breathe(A);
    const frame = emitAeonFrame(A);
    const v = admitAndVerify(frame);
    if (!v.admitted) allAdmitted = false;
    const colorName = frame.aeon.braidColor.padEnd(6);
    console.log(
      `  ${String(frame.breath).padStart(3)}  ${frame.aeon.braidPhase.padEnd(17)} ${colorName} ` +
      `${String(frame.aeon.intensity).padStart(3)}     ${String(frame.aeon.character).padStart(2)}  ` +
      `${frame.aeon.vitality.toFixed(4)}    ${v.recomputedVitality.toFixed(4)}   ${(v.drift ?? 0).toFixed(4)}  ${v.admitted}`,
    );
    breaths++;
  }

  // negative control: a DECORATIVE animation that fakes vitality (pretty number,
  // unchanged witness). The listener MUST reject it — this is the veridical bridge.
  const honest = emitAeonFrame(spawnAeon());
  const fakeFrame = { ...honest, aeon: { ...honest.aeon, vitality: 0.95 } }; // lie
  const fakeVerdict = admitAndVerify(fakeFrame);
  if (!fakeVerdict.admitted && fakeVerdict.reason === 'vitality-faked') faked++;

  // negative control 2: tampered witness (forge digest but keep old hash).
  const tampered = { ...honest, aeon: { ...honest.aeon, psdDigest: honest.aeon.psdDigest.map(() => 1) } };
  const tamperVerdict = admitAndVerify(tampered);

  // negative control 3: protocol69 fails closed.
  const unadmitted = { ...honest, protocol69: 70 };
  const unadmittedVerdict = admitAndVerify(unadmitted);

  console.log('\n--- veridical bridge: decorative-animation rejection (negative controls) ---');
  console.log(`  faked-vitality frame (0.95): admitted=${fakeVerdict.admitted} reason=${fakeVerdict.reason} recomputed=${fakeVerdict.recomputedVitality?.toFixed(4)}`);
  console.log(`  tampered-witness frame:      admitted=${tamperVerdict.admitted} reason=${tamperVerdict.reason}`);
  console.log(`  wrong-protocol69 (70):       admitted=${unadmittedVerdict.admitted} reason=${unadmittedVerdict.reason}`);

  console.log('\n--- resolution (geodesicLength:0 -> channel idle -> $0) ---');
  console.log(`  Aeon resolved after ${A.breath} breaths: intensity=${A.mag} character=${A.sign} vitality=${resolutionOf(A.acc).toFixed(4)}`);
  console.log(`  resolved single mode = bin ${emitAeonFrame(A).aeon.resolvedMode}/512 (one coherent carrier)`);
  console.log(`  every honest frame admitted+verified: ${allAdmitted}; decorative frames rejected: ${faked === 1}`);
  console.log('\n  circulating value at resolution =', A.seed + A.breath, '(seed + sustained-count = the integral of continued existence)');

  return { breaths: A.breath, resolved: isResolved(A), allAdmitted, decorativeRejected: faked === 1 };
}

main();
