// @a0n/gnosis/distributed-inference-host/src/__tests__/amplituhedron-coordinator.test.ts
/**
 * Amplituhedron coordinator loop — host-side fan-out test.
 *
 * Validates `AMPLITUHEDRON_PLAN.md` §4 wired into `Pipeline.generate()`:
 *   1. First call: every station reports MISS → prefill runs → host fires
 *      capture on each station.
 *   2. Second call (same prompt): every station reports HIT → prefill is
 *      SKIPPED entirely; the exit station's cached tail residual is fed
 *      directly into lm_head and decode proceeds.
 *
 * The default-ON gate is asserted by leaving `GNOSIS_AMPLITUHEDRON_COORDINATOR`
 * unset for the run.
 *
 * Mock surface mirrors the optional `amplituhedronReplay` /
 * `amplituhedronCapture` methods that real stations expose against the
 * `/amplituhedron/replay` and `/amplituhedron/capture` endpoints in
 * `apps/distributed-inference-worker/src/index.ts:1880-1970`.
 */

import { describe, expect, it } from 'vitest';

import { mapToMahaShard, resolveMahaShardRoute } from '../amplituhedron-sharding';
import { issueAmplituhedronReplayXorCertificate, Pipeline, prefixHashU64 } from '../pipeline';
import type { IStation } from '../types';

const HIDDEN_DIM = 4;
const VOCAB = 8;

interface Counters {
  forwards: number;
  embeds: number;
  lmHeads: number;
  replays: number;
  captures: number;
}

interface CachedVolume {
  prefixHash: string;
  prefixLen: number;
  tailResidual: Float32Array;
  prefixResidual: Float32Array;
}

function buildAmplituhedronStation(opts: {
  startLayer: number;
  endLayer: number;
  role: 'entry' | 'mid' | 'exit';
  /** Token id this station's lm_head will return (argmax of logits). */
  lmHeadArgmax?: number;
}): { station: IStation; counters: Counters; cache: CachedVolume[] } {
  const counters: Counters = {
    forwards: 0,
    embeds: 0,
    lmHeads: 0,
    replays: 0,
    captures: 0,
  };
  const cache: CachedVolume[] = [];

  // The mock station's "tail residual" is just a deterministic function of
  // the prompt length so the test can check both branches end-to-end.
  const synthesizeTail = (prefixLen: number): Float32Array => {
    const out = new Float32Array(HIDDEN_DIM);
    for (let i = 0; i < HIDDEN_DIM; i++) out[i] = (prefixLen + i) * 0.001;
    return out;
  };
  const synthesizePrefix = (prefixLen: number): Float32Array => {
    const out = new Float32Array(prefixLen * HIDDEN_DIM);
    for (let i = 0; i < out.length; i++) out[i] = (prefixLen + i) * 0.0001;
    return out;
  };

  const station = {
    startLayer: opts.startLayer,
    endLayer: opts.endLayer,
    role: opts.role,
    maxBatchSize: 8,

    async embed({ tokens }: { tokens: Uint32Array; startPos: number; requestId?: string }) {
      counters.embeds++;
      const out = new Float32Array(tokens.length * HIDDEN_DIM);
      for (let i = 0; i < out.length; i++) out[i] = (i + 1) * 0.001;
      return out;
    },

    async forward({
      residual,
      sequenceLength,
    }: {
      residual: Float32Array;
      startPos: number;
      sequenceLength: number;
      requestId?: string;
    }) {
      counters.forwards++;
      const out = new Float32Array(sequenceLength * HIDDEN_DIM);
      for (let i = 0; i < out.length; i++) out[i] = (residual[i] ?? 0) + 0.01;
      return { residual: out };
    },

    async lmHead({ sequenceLength }: { residual: Float32Array; sequenceLength: number; requestId?: string }) {
      counters.lmHeads++;
      const out = new Float32Array(VOCAB);
      out[opts.lmHeadArgmax ?? 3] = 1;
      // Suppress unused-arg warning under noUnusedParameters.
      void sequenceLength;
      return out;
    },

    async cobordismEmbed() {
      throw new Error('not used');
    },
    async cobordismLmHead() {
      throw new Error('not used');
    },
    async bizarroForward() {
      return null;
    },

    // Optional Amplituhedron surface — duck-typed by Pipeline.
    async amplituhedronReplay({
      prefixHash,
      prefixLen,
    }: {
      prefixHash: string;
      prefixLen: number;
      layerLo: number;
      layerHi: number;
      requestId?: string;
    }) {
      counters.replays++;
      const hit = cache.find(
        (v) => v.prefixHash === prefixHash && v.prefixLen === prefixLen,
      );
      if (!hit) return { hit: false };
      return {
        hit: true,
        tailResidual: hit.tailResidual,
        prefixResidual: hit.prefixResidual,
      };
    },

    async amplituhedronCapture({
      prefixHash,
      prefixLen,
      tailResidual,
    }: {
      prefixHash: string;
      prefixLen: number;
      layerLo: number;
      layerHi: number;
      tailResidual?: Float32Array;
      requestId?: string;
    }) {
      counters.captures++;
      const capturedPrefix = tailResidual ?? synthesizePrefix(prefixLen);
      const capturedTail =
        capturedPrefix.length === prefixLen * HIDDEN_DIM
          ? capturedPrefix.slice(capturedPrefix.length - HIDDEN_DIM)
          : capturedPrefix;
      cache.push({
        prefixHash,
        prefixLen,
        tailResidual: capturedTail.length > 0 ? capturedTail : synthesizeTail(prefixLen),
        prefixResidual: capturedPrefix,
      });
    },
  } as unknown as IStation;

  return { station, counters, cache };
}

function makePipeline(stations: IStation[]): Pipeline {
  return new Pipeline({
    stations,
    hiddenDim: HIDDEN_DIM,
    vocabSize: VOCAB,
    stopTokens: [],
  });
}

describe('Pipeline Amplituhedron coordinator (Death #2)', () => {
  it('hashes Uint32Array views without changing prefix identity', async () => {
    const base = new Uint32Array([99, 7, 11, 13, 17, 101]);
    const copied = new Uint32Array([7, 11, 13, 17]);
    const view = base.subarray(1, 5);

    await expect(prefixHashU64(view)).resolves.toBe(await prefixHashU64(copied));
  });

  it('maps decimal prefix hashes to the same Maha shard as BigInt', () => {
    const samples = [
      0n,
      1n,
      20n,
      21n,
      199n,
      18446744073709551615n,
    ];

    for (const sample of samples) {
      expect(mapToMahaShard(sample.toString())).toBe(mapToMahaShard(sample));
      expect(resolveMahaShardRoute(21, sample.toString())).toEqual(
        resolveMahaShardRoute(21, sample),
      );
    }
  });

  it('projects replay misses into a station-level Grassmannian XOR certificate', () => {
    const certificate = issueAmplituhedronReplayXorCertificate({
      prefixHash: '42',
      prefixLen: 7,
      stationHits: [true, false, true, false],
    });

    expect(certificate.certificateId).toBe('grassmannian-xor-round:42:7:replay');
    expect(certificate.status).toBe('failed');
    expect(certificate.defectRank).toBe(2);
    expect(certificate.parityClass).toBe(
      '1:distributed-inference-amplituhedron:plucker-lane-1 xor 3:distributed-inference-amplituhedron:plucker-lane-3',
    );
  });

  it('marks a full replay hit as a balanced Grassmannian XOR round', () => {
    const certificate = issueAmplituhedronReplayXorCertificate({
      prefixHash: '42',
      prefixLen: 7,
      stationHits: [true, true, true, true],
    });

    expect(certificate.status).toBe('measured');
    expect(certificate.balanced).toBe(true);
    expect(certificate.defectRank).toBe(0);
    expect(certificate.parityClass).toBe('0');
  });

  it('skips prefill on the second generate() when both stations report a HIT', async () => {
    const s0 = buildAmplituhedronStation({
      startLayer: 0,
      endLayer: 4,
      role: 'entry',
    });
    const s1 = buildAmplituhedronStation({
      startLayer: 4,
      endLayer: 8,
      role: 'exit',
      lmHeadArgmax: 5,
    });
    const pipe = makePipeline([s0.station, s1.station]);

    const promptTokens = new Uint32Array([7, 11, 13, 17, 19]);

    // ── Call 1: cache cold. Both stations MISS → prefill runs, capture
    //   fires fire-and-forget on each station.
    const out1 = await pipe.generate({ promptTokens, maxNewTokens: 1 });
    expect(out1.length).toBe(promptTokens.length + 1);

    // Both stations got replay calls (the miss probe).
    expect(s0.counters.replays).toBe(1);
    expect(s1.counters.replays).toBe(1);

    // Prefill ran: embed (1) + forward calls (>=1) on each station.
    expect(s0.counters.embeds).toBe(1);
    expect(s0.counters.forwards).toBeGreaterThan(0);
    expect(s1.counters.forwards).toBeGreaterThan(0);
    expect(s1.counters.lmHeads).toBeGreaterThanOrEqual(1);

    // Capture fired on both stations (fire-and-forget — wait one
    // microtask flush so allSettled has settled).
    await new Promise((r) => setTimeout(r, 0));
    expect(s0.counters.captures).toBe(1);
    expect(s1.counters.captures).toBe(1);
    expect(s0.cache.length).toBe(1);
    expect(s1.cache.length).toBe(1);
    expect(s0.cache[0]!.prefixResidual.length).toBe(promptTokens.length * HIDDEN_DIM);
    expect(s1.cache[0]!.prefixResidual.length).toBe(promptTokens.length * HIDDEN_DIM);

    // Snapshot pre-call-2 counters so we can assert NO new prefill work.
    const forwardsBefore = s0.counters.forwards + s1.counters.forwards;
    const embedsBefore = s0.counters.embeds + s1.counters.embeds;
    const lmHeadsBefore = s1.counters.lmHeads;

    // ── Call 2: same prompt. Both stations HIT → prefill SKIPPED. The
    //   coordinator goes straight to lm_head on the exit station's cached
    //   tail residual, then runs the decode step.
    const out2 = await pipe.generate({ promptTokens, maxNewTokens: 1 });
    expect(out2.length).toBe(promptTokens.length + 1);

    // Both stations were probed once more.
    expect(s0.counters.replays).toBe(2);
    expect(s1.counters.replays).toBe(2);

    // CRITICAL: prefill did NOT run on the prompt chain on station 0.
    // The host skipped embed entirely and only the EXIT lm_head fired
    // (plus the per-token decode work, which is the same forward count
    // both branches incur for maxNewTokens=1).
    expect(s0.counters.embeds).toBe(embedsBefore); // no new prompt embed
    // Decode step still walks every station, so we expect exactly +1
    // forward per station from the single decoded token, NOT the full
    // multi-chunk prefill traffic.
    expect(s0.counters.forwards - (forwardsBefore - s1.counters.forwards)).toBeLessThanOrEqual(
      forwardsBefore, // sanity: not zero on s0 if decode runs through it
    );
    // The exit station's lm_head was called at least once more for the
    // skipped-prefill path (the decode step uses sample-next or lm_head
    // again).
    expect(s1.counters.lmHeads).toBeGreaterThan(lmHeadsBefore);
  });

  it('recomputes only the replay defect suffix when a cached prefix exports full residuals', async () => {
    const s0 = buildAmplituhedronStation({
      startLayer: 0,
      endLayer: 2,
      role: 'entry',
    });
    const s1 = buildAmplituhedronStation({
      startLayer: 2,
      endLayer: 4,
      role: 'mid',
    });
    const s2 = buildAmplituhedronStation({
      startLayer: 4,
      endLayer: 6,
      role: 'exit',
      lmHeadArgmax: 6,
    });
    const pipe = makePipeline([s0.station, s1.station, s2.station]);
    const promptTokens = new Uint32Array([3, 5, 8]);

    await pipe.generate({ promptTokens, maxNewTokens: 0 });
    await new Promise((r) => setTimeout(r, 0));
    expect(s0.cache.length).toBe(1);
    expect(s1.cache.length).toBe(1);
    expect(s2.cache.length).toBe(1);

    s1.cache.length = 0;
    s2.cache.length = 0;

    const s0EmbedsBefore = s0.counters.embeds;
    const s0ForwardsBefore = s0.counters.forwards;
    const s1ForwardsBefore = s1.counters.forwards;
    const s2ForwardsBefore = s2.counters.forwards;
    const s2LmHeadsBefore = s2.counters.lmHeads;

    await pipe.generate({ promptTokens, maxNewTokens: 0 });

    expect(s0.counters.embeds).toBe(s0EmbedsBefore);
    expect(s0.counters.forwards).toBe(s0ForwardsBefore);
    expect(s1.counters.forwards).toBe(s1ForwardsBefore + 1);
    expect(s2.counters.forwards).toBe(s2ForwardsBefore + 1);
    expect(s2.counters.lmHeads).toBe(s2LmHeadsBefore);
  });

  it('probes and captures only the Maha shard for equivalent cache replicas', async () => {
    const stations = Array.from({ length: 21 }, (_, index) =>
      buildAmplituhedronStation({
        startLayer: 0,
        endLayer: 24,
        role: index === 0 ? 'entry' : index === 20 ? 'exit' : 'mid',
      }),
    );
    const pipe = new Pipeline({
      stations: stations.map((s) => s.station),
      hiddenDim: HIDDEN_DIM,
      vocabSize: VOCAB,
      stopTokens: [],
      amplituhedronSharding: true,
    });

    await pipe.generate({
      promptTokens: new Uint32Array([7, 11, 13, 17, 19]),
      maxNewTokens: 0,
    });
    await new Promise((r) => setTimeout(r, 0));

    const replayCount = stations.reduce((sum, s) => sum + s.counters.replays, 0);
    const captureCount = stations.reduce((sum, s) => sum + s.counters.captures, 0);

    expect(replayCount).toBe(1);
    expect(captureCount).toBe(1);
  });

  it('keeps full layer coverage when sharding is enabled on a sequential mesh', async () => {
    const stations = Array.from({ length: 21 }, (_, index) =>
      buildAmplituhedronStation({
        startLayer: index,
        endLayer: index + 1,
        role: index === 0 ? 'entry' : index === 20 ? 'exit' : 'mid',
      }),
    );
    const pipe = new Pipeline({
      stations: stations.map((s) => s.station),
      hiddenDim: HIDDEN_DIM,
      vocabSize: VOCAB,
      stopTokens: [],
      amplituhedronSharding: true,
    });

    await pipe.generate({
      promptTokens: new Uint32Array([7, 11, 13, 17, 19]),
      maxNewTokens: 0,
    });
    await new Promise((r) => setTimeout(r, 0));

    const replayCount = stations.reduce((sum, s) => sum + s.counters.replays, 0);
    const captureCount = stations.reduce((sum, s) => sum + s.counters.captures, 0);

    expect(replayCount).toBe(21);
    expect(captureCount).toBe(21);
  });
});
