// @a0n/.ignored_gnosis/src/esc/__tests__/edge-side-compute.fp48.test.ts
/**
 * Flag-gated fp48 port of the ESC cache keying surface.
 *
 * Covers:
 * - `ESC_FP48_ENABLED` off → keys are byte-identical to the fp64-v1 baseline.
 * - `ESC_FP48_ENABLED` on  → digest is `cacheFingerprint48HexSync` (12 hex).
 * - fp48 key carries the `gnosis:esc:fp48-v1` prefix via classification.
 * - Zero-marginal-fold on cache hit survives the fp48 switch
 *   (Lean: `ForkRaceFoldTheorems.CacheObservation`).
 *
 * Provenance of the 4:3 shrink:
 *   `open-source/gnosis/lean/Lean/ForkRaceFoldTheorems/BizarroCompiler.lean`
 *   — `fp48Compressor`, `bitwise_shrink_4_3`, `fp48_concrete_shrink`,
 *   `fp48_concrete_4x`.
 */

import { describe, expect, it } from '@a0n/gnosis/test';
import {
  cacheFingerprint48HexSync,
  FP48_HEX_LEN,
} from '@a0n/bitwise/cache-fp48';
import { alchemicHash48HexSync } from '@a0n/bitwise/cache-fp64';
import {
  canonicalEscInput,
  computeEscCacheKey,
  EscMemoryTier,
  slicePackedKnot,
  type EscWasmSegmentRecord,
} from '../edge-side-compute';
import {
  classifyEscCacheLookup,
  marginalFoldWork,
  positiveFoldWork,
} from '../cache-observation';
import { dnaToKnot } from '../../runtime/packed-knot';

const ENV_FLAG = 'ESC_FP48_ENABLED';

function withFlag<T>(value: string | undefined, body: () => T): T {
  const proc = (globalThis as { process?: { env?: Record<string, string | undefined> } }).process;
  if (!proc || !proc.env) {
    // No Node-shaped env; exercise body as-is (flag is effectively off).
    return body();
  }
  const prior = proc.env[ENV_FLAG];
  if (value === undefined) {
    delete proc.env[ENV_FLAG];
  } else {
    proc.env[ENV_FLAG] = value;
  }
  try {
    return body();
  } finally {
    if (prior === undefined) {
      delete proc.env[ENV_FLAG];
    } else {
      proc.env[ENV_FLAG] = prior;
    }
  }
}

function sampleInput(): Parameters<typeof computeEscCacheKey>[0] {
  const knot = dnaToKnot('ACGTACGTACGT');
  const slice = slicePackedKnot(knot, 0, 8);
  return {
    knotSlice: slice,
    entryCrossingIndex: 0,
    canonicalInput: canonicalEscInput({ n: 1, tag: 'fp48' }),
    artifactFingerprint: 'artifact:sha256:abc',
    runtimeBindingHash: 'binding:wasm:v1',
    policySalt: 'tenant:test',
    handlerVersionParts: ['main', 'json-v1'],
  };
}

describe('edge-side-compute fp48 port', () => {
  it('flag off yields byte-identical fp64-v1 baseline keys', () => {
    const input = sampleInput();
    const baseline = withFlag(undefined, () => computeEscCacheKey(input));
    const withFalse = withFlag('false', () => computeEscCacheKey(input));
    const withZero = withFlag('0', () => computeEscCacheKey(input));
    const withEmpty = withFlag('', () => computeEscCacheKey(input));
    expect(withFalse).toBe(baseline);
    expect(withZero).toBe(baseline);
    expect(withEmpty).toBe(baseline);
    // Baseline is the existing fp64-v1 digest surface ('a0n' prefix + 12 hex).
    expect(baseline.startsWith('a0n')).toBe(true);
    expect(baseline.length).toBe('a0n'.length + 12);
  });

  it('flag on selects the fp48 digest (pure 12 lowercase hex)', () => {
    const input = sampleInput();
    const fp48Key = withFlag('1', () => computeEscCacheKey(input));
    expect(fp48Key.length).toBe(FP48_HEX_LEN);
    expect(fp48Key.length).toBe(12);
    expect(/^[0-9a-f]{12}$/.test(fp48Key)).toBe(true);
    // fp48 emits no 'a0n' prefix; 12 hex chars is the whole key.
    expect(fp48Key.startsWith('a0n')).toBe(false);
  });

  it('truthy flag variants all enable fp48', () => {
    const input = sampleInput();
    const baseKey = withFlag('1', () => computeEscCacheKey(input));
    for (const v of ['1', 'true', 'TRUE', 'yes', 'on']) {
      const k = withFlag(v, () => computeEscCacheKey(input));
      expect(k).toBe(baseKey);
      expect(k.length).toBe(12);
    }
  });

  it('fp48 key is strictly shorter than fp64-v1 key on the same input', () => {
    const input = sampleInput();
    const fp64Key = withFlag(undefined, () => computeEscCacheKey(input));
    const fp48Key = withFlag('1', () => computeEscCacheKey(input));
    // fp64-v1 surface: 'a0n' + 12 hex = 15 chars. fp48-v1 surface: 12 hex.
    // Either way, fp48 saves bytes on every key.
    expect(fp48Key.length).toBeLessThan(fp64Key.length);
    expect(fp64Key.length - fp48Key.length).toBeGreaterThanOrEqual(3);
  });

  it('fp48 prefix tag resolves from the digest surface exposed by cache-fp48', () => {
    // The ESC key itself is opaque 12 hex; the version tag lives in the fold
    // prefix that feeds the digest. Confirm that the fp48 digest matches what
    // `cacheFingerprint48HexSync` produces over the *same* concatenated
    // material that includes the 'gnosis:esc:fp48-v1\0' prefix.
    const input = sampleInput();
    const fp48Key = withFlag('1', () => computeEscCacheKey(input));
    // Reconstruct the folded material. We don't export the internal concat,
    // but we can at least witness that the fp48 digest differs from the
    // fp64-v1 digest (distinct prefix feeds distinct bytes feed distinct hash).
    const fp64Key = withFlag(undefined, () => computeEscCacheKey(input));
    expect(fp48Key).not.toBe(fp64Key);
    // And that the fp48 digest lies in the 12-hex fp48 surface.
    const probe = cacheFingerprint48HexSync('probe');
    expect(probe.length).toBe(FP48_HEX_LEN);
    expect(/^[0-9a-f]{12}$/.test(probe)).toBe(true);
    // Defensive: the fp64 surface still emits its 'a0n'-tagged 15-char form,
    // so we are definitively on distinct digest channels.
    expect(alchemicHash48HexSync(new TextEncoder().encode('probe')).startsWith('a0n')).toBe(true);
  });

  it('cache hits observe zero-marginal-fold with fp48 keys', async () => {
    const tier = new EscMemoryTier();
    const input = sampleInput();
    const key = withFlag('1', () => computeEscCacheKey(input));
    const record: EscWasmSegmentRecord = {
      version: 1,
      artifactId: 'artifact:sha256:abc',
      exportName: 'main',
      abi: 'json-v1',
      payload: { ok: true },
      logs: '',
    };
    await tier.set(key, record);

    const cached = await tier.get(key);
    const resolution = classifyEscCacheLookup({
      cached,
      signatureMatches: cached !== null && cached.artifactId === record.artifactId,
      entryValid: true,
      w: positiveFoldWork(1),
    });
    expect(resolution.kind).toBe('hit');
    expect(marginalFoldWork(resolution)).toBe(0);
  });

  it('stale fp48 entries still report positive marginal fold work', async () => {
    const tier = new EscMemoryTier();
    const input = sampleInput();
    const key = withFlag('1', () => computeEscCacheKey(input));
    const record: EscWasmSegmentRecord = {
      version: 1,
      artifactId: 'artifact:sha256:abc',
      exportName: 'main',
      abi: 'json-v1',
      payload: { ok: true },
      logs: '',
    };
    await tier.set(key, record);
    const cached = await tier.get(key);
    const resolution = classifyEscCacheLookup({
      cached,
      signatureMatches: cached !== null && cached.artifactId === record.artifactId,
      entryValid: false,
      w: positiveFoldWork(3),
    });
    expect(resolution.kind).toBe('stale_miss');
    expect(marginalFoldWork(resolution)).toBe(3);
  });

  it('fp48 keys remain stable across repeated calls with identical inputs', () => {
    const input = sampleInput();
    const a = withFlag('1', () => computeEscCacheKey(input));
    const b = withFlag('1', () => computeEscCacheKey(input));
    expect(a).toBe(b);
  });

  it('fp48 keys diverge when the entry crossing index changes', () => {
    const base = sampleInput();
    const a = withFlag('1', () => computeEscCacheKey(base));
    const b = withFlag('1', () =>
      computeEscCacheKey({ ...base, entryCrossingIndex: 1 })
    );
    expect(a).not.toBe(b);
  });
});
