// @a0n/.ignored_gnosis/src/adt.ts
/**
 * adt.ts -- Algebraic Data Types for gnosis topologies
 *
 * Sum types for closed state modeling. Product types for structured values.
 * Exhaustive matching when branching on closed variants.
 *
 * Roadmap item #2: dependency gate for Option/Result and destructuring.
 */

// ============================================================================
// Variant -- a single case of a sum type
// ============================================================================

export interface Variant<T = unknown> {
  /** Tag identifying which case this is */
  readonly tag: string;
  /** Payload carried by this variant */
  readonly value: T;
}

export function variant<T>(tag: string, value: T): Variant<T> {
  return { tag, value };
}

// ============================================================================
// SumType -- closed union of tagged variants
// ============================================================================

export interface SumTypeDefinition {
  /** Name of the sum type */
  readonly name: string;
  /** All valid case tags */
  readonly cases: readonly string[];
  /** Whether the type is sealed (no extension) */
  readonly sealed: boolean;
}

/**
 * Handles the gnosis define Sum Type workflow.
 */
export function defineSumType(
  name: string,
  cases: readonly string[],
  sealed: boolean = true
): SumTypeDefinition {
  return { name, cases, sealed };
}

/** Built-in sum types */
export const BUILTIN_SUM_TYPES: Record<string, SumTypeDefinition> = {
  Result: defineSumType('Result', ['ok', 'err']),
  Option: defineSumType('Option', ['some', 'none']),
  Either: defineSumType('Either', ['left', 'right']),
  Ordering: defineSumType('Ordering', ['lt', 'eq', 'gt']),
  TaskStatus: defineSumType('TaskStatus', [
    'pending',
    'running',
    'success',
    'error',
    'timeout',
    'cancelled',
    'vented',
  ]),
  ConvergenceState: defineSumType('ConvergenceState', [
    'converged',
    'diverged',
    'oscillating',
    'stalled',
  ]),
};

// ============================================================================
// ProductType -- named fields
// ============================================================================

export interface FieldDefinition {
  readonly name: string;
  readonly required: boolean;
  readonly defaultValue?: unknown;
}

export interface ProductTypeDefinition {
  readonly name: string;
  readonly fields: readonly FieldDefinition[];
}

/**
 * Handles the gnosis define Product Type workflow.
 */
export function defineProductType(
  name: string,
  fields: readonly FieldDefinition[]
): ProductTypeDefinition {
  return { name, fields };
}

/**
 * Creates the Record.
 */
export function createRecord(
  typeDef: ProductTypeDefinition,
  values: Record<string, unknown>
): Record<string, unknown> {
  try {
    const record: Record<string, unknown> = {};
    for (const field of typeDef.fields) {
      if (field.name in values) {
        record[field.name] = values[field.name];
      } else if (field.defaultValue !== undefined) {
        record[field.name] = field.defaultValue;
      } else if (field.required) {
        throw new Error(
          `Missing required field '${field.name}' in ${typeDef.name}`
        );
      }
    }
    return record;

  } catch (error) {
    throw new Error(`[createRecord] ${error instanceof Error ? error.message : String(error)}`);
  }
}

// ============================================================================
// Pattern matching -- exhaustive match on sum types
// ============================================================================

export interface MatchArm<T, R> {
  readonly tag: string;
  readonly handler: (value: T) => R;
}

export interface MatchResult<R> {
  readonly matched: boolean;
  readonly tag: string;
  readonly value: R | undefined;
}

/**
 * Exhaustive match on a variant against a sum type definition.
 * Throws if any case is unhandled.
 */
export function matchExhaustive<T, R>(
  typeDef: SumTypeDefinition,
  value: Variant<T>,
  arms: Record<string, (v: T) => R>
): R {
  try {
    // Check exhaustiveness
    const missing = typeDef.cases.filter((c) => !(c in arms));
    if (missing.length > 0) {
      throw new ExhaustivenessError(typeDef.name, missing);
    }

    const handler = arms[value.tag];
    if (!handler) {
      throw new Error(`Unknown variant '${value.tag}' for type ${typeDef.name}`);
    }

    return handler(value.value);

  } catch (error) {
    throw new Error(`[matchExhaustive] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Non-exhaustive match with a default fallback.
 */
export function matchPartial<T, R>(
  value: Variant<T>,
  arms: Record<string, (v: T) => R>,
  fallback: (v: Variant<T>) => R
): R {
  const handler = arms[value.tag];
  return handler ? handler(value.value) : fallback(value);
}

// ============================================================================
// Exhaustiveness checking -- compile-time-like verification at runtime
// ============================================================================

export class ExhaustivenessError extends Error {
  readonly typeName: string;
  readonly missingCases: string[];

  constructor(typeName: string, missingCases: string[]) {
    super(
      `Non-exhaustive match on '${typeName}': missing cases [${missingCases.join(
        ', '
      )}]`
    );
    this.name = 'ExhaustivenessError';
    this.typeName = typeName;
    this.missingCases = missingCases;
  }
}

/**
 * Check whether a set of handled cases covers all variants.
 */
export function checkExhaustiveness(
  typeDef: SumTypeDefinition,
  handledCases: string[]
): { exhaustive: boolean; missing: string[]; extra: string[] } {
  const handled = new Set(handledCases);
  const missing = typeDef.cases.filter((c) => !handled.has(c));
  const extra = handledCases.filter((c) => !typeDef.cases.includes(c));
  return { exhaustive: missing.length === 0, missing, extra };
}

/**
 * Validate that outgoing edges from a node cover all cases of its sum type.
 * This is the topology-level exhaustiveness check.
 */
export function validateNodeExhaustiveness(
  nodeId: string,
  typeDef: SumTypeDefinition,
  edgeCases: string[]
): { valid: boolean; missing: string[]; diagnostic: string | null } {
  try {
    const { exhaustive, missing } = checkExhaustiveness(typeDef, edgeCases);
    if (exhaustive) {
      return { valid: true, missing: [], diagnostic: null };
    }
    return {
      valid: false,
      missing,
      diagnostic: `Node '${nodeId}' (${
        typeDef.name
      }): missing cases [${missing.join(', ')}]`,
    };

  } catch (error) {
    throw new Error(`[validateNodeExhaustiveness] ${error instanceof Error ? error.message : String(error)}`);
  }
}

// ============================================================================
// Type registry -- custom sum and product types
// ============================================================================

export class TypeRegistry {
  private sumTypes = new Map<string, SumTypeDefinition>();
  private productTypes = new Map<string, ProductTypeDefinition>();

  constructor() {
    // Register builtins
    for (const [name, def] of Object.entries(BUILTIN_SUM_TYPES)) {
      this.sumTypes.set(name, def);
    }
  }

  registerSumType(def: SumTypeDefinition): void {
    try {
      if (this.sumTypes.has(def.name)) {
        throw new Error(`Sum type '${def.name}' already registered`);
      }
      this.sumTypes.set(def.name, def);
  
    } catch (error) {
      throw new Error(`[registerSumType] ${error instanceof Error ? error.message : String(error)}`);
    }
}

  registerProductType(def: ProductTypeDefinition): void {
    try {
      if (this.productTypes.has(def.name)) {
        throw new Error(`Product type '${def.name}' already registered`);
      }
      this.productTypes.set(def.name, def);
  
    } catch (error) {
      throw new Error(`[registerProductType] ${error instanceof Error ? error.message : String(error)}`);
    }
}

  getSumType(name: string): SumTypeDefinition | undefined {
    return this.sumTypes.get(name);
  }

  getProductType(name: string): ProductTypeDefinition | undefined {
    return this.productTypes.get(name);
  }

  hasSumType(name: string): boolean {
    return this.sumTypes.has(name);
  }

  allSumTypes(): SumTypeDefinition[] {
    return [...this.sumTypes.values()];
  }

  allProductTypes(): ProductTypeDefinition[] {
    return [...this.productTypes.values()];
  }
}
