// @a0n/.ignored_gnosis/src/advisory-allocation-kernel.ts
import {
  CERTIFIED_ALLOCATION_KERNEL_THEOREM_IDS,
  certifiedAllocationStep,
  confidenceFloorSatisfied,
  type CertifiedAllocationDecision,
  type CertifiedCalibrationPlan,
  type CertifiedCalibrationPointLedger,
  type CertifiedLoopBudget,
  type CertifiedObservationFrame,
  type NormalizedMilliPoint,
  type ObservationZone,
} from './certified-allocation-kernel';

export const ADVISORY_ALLOCATION_KERNEL_THEOREM_IDS = [
  'THM-ADVISORY-ALLOCATION-DECAY-EXACT',
  'THM-ADVISORY-ALLOCATION-STABILITY-BOUNDARY',
  'THM-ADVISORY-ALLOCATION-COMMIT-POINT-FAIL-CLOSED',
  'THM-ADVISORY-ALLOCATION-RISK-LIMIT-FAIL-CLOSED',
  'THM-ADVISORY-ALLOCATION-STALE-OBSERVATION-FAIL-CLOSED',
  'THM-ADVISORY-ALLOCATION-CONFIDENCE-DECAY-FAIL-CLOSED',
  'THM-ADVISORY-ALLOCATION-ADVERSE-SELECTION-FAIL-CLOSED',
  'THM-ADVISORY-ALLOCATION-HARD-DEADLINE-FAIL-CLOSED',
  'THM-ADVISORY-ALLOCATION-CERTIFIED-BRIDGE',
  'THM-ADVISORY-ALLOCATION-LOOP-PROGRESS-OR-EXPLICIT-FAILURE',
] as const;

export type AdvisoryAllocationKernelTheoremId =
  (typeof ADVISORY_ALLOCATION_KERNEL_THEOREM_IDS)[number];

export interface AdvisoryObservationFrame {
  readonly observation: CertifiedObservationFrame;
  readonly currentPrice: number;
  readonly lastPrice: number;
  readonly displayedDepth: number;
  readonly elapsedSinceUpdateMs: number;
  readonly maxObservationAgeMs: number;
  readonly decayPerMs: number;
  readonly priceMoveThreshold: number;
}

export interface RiskEnvelope {
  readonly inventoryExposure: number;
  readonly inventoryLimit: number;
  readonly openOrderCount: number;
  readonly openOrderLimit: number;
}

export interface AdvisoryBudget {
  readonly deadlineMs: number;
  readonly elapsedMs: number;
  readonly commitNumerator: number;
  readonly commitDenominator: number;
}

export type AdvisorySafeStateReason =
  | 'deadline-exceeded'
  | 'not-ready'
  | 'stale-observation'
  | 'adverse-selection'
  | 'risk-limit-reached'
  | 'no-candidates';

export type AdvisoryAllocationDecision =
  | {
      readonly kind: 'allocate';
      readonly candidateId: number;
    }
  | {
      readonly kind: 'safe-state';
      readonly reason: AdvisorySafeStateReason;
    };

export interface AdvisoryAllocationKernelInput {
  readonly generatedAt?: string;
  readonly source?: string;
  readonly plan: CertifiedCalibrationPlan;
  readonly points: readonly CertifiedCalibrationPointLedger[];
  readonly modelSolved: boolean;
  readonly ingestLocked: boolean;
  readonly frame: AdvisoryObservationFrame;
  readonly candidateIds: readonly number[];
  readonly risk: RiskEnvelope;
  readonly budget: AdvisoryBudget;
}

export interface AdvisoryAllocationKernelAggregate {
  readonly priceMove: number;
  readonly confidenceDecay: number;
  readonly decayedConfidenceScore: number;
  readonly observationFresh: boolean;
  readonly stabilityWitness: boolean;
  readonly riskWithinLimits: boolean;
  readonly marketReady: boolean;
  readonly withinBudget: boolean;
  readonly commitPointReached: boolean;
  readonly confidenceFloorSatisfied: boolean;
  readonly certifiedBridgeDecision: CertifiedAllocationDecision;
  readonly inheritedCertifiedTheoremIds: readonly string[];
}

export interface AdvisoryAllocationKernelBundle {
  readonly generatedAt: string;
  readonly source: string;
  readonly theoremIds: readonly AdvisoryAllocationKernelTheoremId[];
  readonly decision: AdvisoryAllocationDecision;
  readonly aggregate: AdvisoryAllocationKernelAggregate;
}

const SYNTHETIC_VISIBLE_POINT: NormalizedMilliPoint = { x: 500, y: 500 };

function nowIso(): string {
  return new Date().toISOString();
}

function clampNonNegativeInt(value: number): number {
  try {
    if (!Number.isFinite(value)) {
      return 0;
    }
    return Math.max(0, Math.floor(value));

  } catch (error) {
    throw new Error(`[clampNonNegativeInt] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function normalizeObservationZone(zone: ObservationZone): ObservationZone {
  return zone;
}

function normalizeObservation(
  observation: CertifiedObservationFrame
): CertifiedObservationFrame {
  const gazePoint = observation.gazePoint
    ? {
        x: clampNonNegativeInt(observation.gazePoint.x),
        y: clampNonNegativeInt(observation.gazePoint.y),
      }
    : null;

  return {
    gazePoint,
    zone: normalizeObservationZone(observation.zone),
    confidenceScore: clampNonNegativeInt(observation.confidenceScore),
    confidenceFloor: clampNonNegativeInt(observation.confidenceFloor),
  };
}

function normalizeFrame(frame: AdvisoryObservationFrame): AdvisoryObservationFrame {
  return {
    observation: normalizeObservation(frame.observation),
    currentPrice: clampNonNegativeInt(frame.currentPrice),
    lastPrice: clampNonNegativeInt(frame.lastPrice),
    displayedDepth: clampNonNegativeInt(frame.displayedDepth),
    elapsedSinceUpdateMs: clampNonNegativeInt(frame.elapsedSinceUpdateMs),
    maxObservationAgeMs: clampNonNegativeInt(frame.maxObservationAgeMs),
    decayPerMs: clampNonNegativeInt(frame.decayPerMs),
    priceMoveThreshold: clampNonNegativeInt(frame.priceMoveThreshold),
  };
}

function normalizeRisk(risk: RiskEnvelope): RiskEnvelope {
  return {
    inventoryExposure: clampNonNegativeInt(risk.inventoryExposure),
    inventoryLimit: clampNonNegativeInt(risk.inventoryLimit),
    openOrderCount: clampNonNegativeInt(risk.openOrderCount),
    openOrderLimit: clampNonNegativeInt(risk.openOrderLimit),
  };
}

function normalizeBudget(budget: AdvisoryBudget): AdvisoryBudget {
  return {
    deadlineMs: clampNonNegativeInt(budget.deadlineMs),
    elapsedMs: clampNonNegativeInt(budget.elapsedMs),
    commitNumerator: clampNonNegativeInt(budget.commitNumerator),
    commitDenominator: clampNonNegativeInt(budget.commitDenominator),
  };
}

function normalizeCandidates(candidateIds: readonly number[]): number[] {
  return candidateIds.map((candidateId) => clampNonNegativeInt(candidateId));
}

/**
 * Handles the gnosis price Move workflow.
 */
export function priceMove(frame: AdvisoryObservationFrame): number {
  return Math.abs(frame.currentPrice - frame.lastPrice);
}

/**
 * Handles the gnosis confidence Decay workflow.
 */
export function confidenceDecay(frame: AdvisoryObservationFrame): number {
  return frame.elapsedSinceUpdateMs * frame.decayPerMs;
}

/**
 * Handles the gnosis decayed Confidence Score workflow.
 */
export function decayedConfidenceScore(frame: AdvisoryObservationFrame): number {
  return Math.max(0, frame.observation.confidenceScore - confidenceDecay(frame));
}

/**
 * Handles the gnosis decayed Observation Frame workflow.
 */
export function decayedObservationFrame(
  frame: AdvisoryObservationFrame
): CertifiedObservationFrame {
  const normalizedObservation = normalizeObservation(frame.observation);
  const zone =
    normalizedObservation.gazePoint === null ? 'unavailable' : normalizedObservation.zone;

  return {
    ...normalizedObservation,
    zone,
    confidenceScore: decayedConfidenceScore(frame),
  };
}

/**
 * Handles the gnosis observation Fresh workflow.
 */
export function observationFresh(frame: AdvisoryObservationFrame): boolean {
  return frame.elapsedSinceUpdateMs <= frame.maxObservationAgeMs;
}

/**
 * Handles the gnosis stability Witness workflow.
 */
export function stabilityWitness(frame: AdvisoryObservationFrame): boolean {
  try {
    if (priceMove(frame) > frame.priceMoveThreshold) {
      return frame.displayedDepth > 0;
    }
    return true;

  } catch (error) {
    throw new Error(`[stabilityWitness] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Handles the gnosis risk Within Limits workflow.
 */
export function riskWithinLimits(risk: RiskEnvelope): boolean {
  return (
    risk.inventoryExposure <= risk.inventoryLimit &&
    risk.openOrderCount <= risk.openOrderLimit
  );
}

/**
 * Handles the gnosis market Ready workflow.
 */
export function marketReady(
  plan: CertifiedCalibrationPlan,
  points: readonly CertifiedCalibrationPointLedger[],
  modelSolved: boolean,
  ingestLocked: boolean
): boolean {
  const totalAccepted = points.reduce((sum, point) => sum + point.accepted, 0);
  const required = plan.pointIds.length * Math.max(0, plan.samplesPerPoint);
  const terminated = plan.pointIds.every((pointId) => {
    const point =
      points.find((candidate) => candidate.pointId === pointId) ?? {
        pointId,
        accepted: 0,
        rejected: 0,
        observed: 0,
      };
    return (
      point.accepted >= plan.samplesPerPoint ||
      point.observed >= plan.maxAttemptsPerPoint
    );
  });

  return terminated && totalAccepted >= required && modelSolved && ingestLocked;
}

/**
 * Handles the gnosis advisory Loop Budget workflow.
 */
export function advisoryLoopBudget(budget: AdvisoryBudget): CertifiedLoopBudget {
  return {
    deadlineMs: budget.deadlineMs,
    elapsedMs: budget.elapsedMs,
  };
}

/**
 * Handles the gnosis within Advisory Budget workflow.
 */
export function withinAdvisoryBudget(budget: AdvisoryBudget): boolean {
  return budget.elapsedMs <= budget.deadlineMs;
}

/**
 * Handles the gnosis commit Point Reached workflow.
 */
export function commitPointReached(budget: AdvisoryBudget): boolean {
  try {
    if (budget.commitDenominator === 0) {
      return false;
    }

    return (
      budget.commitDenominator * budget.elapsedMs >=
      budget.commitNumerator * budget.deadlineMs
    );

  } catch (error) {
    throw new Error(`[commitPointReached] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Handles the gnosis advisory Decision Of Certified workflow.
 */
export function advisoryDecisionOfCertified(
  decision: CertifiedAllocationDecision
): AdvisoryAllocationDecision {
  try {
    if (decision.kind === 'allocate') {
      return decision;
    }

    switch (decision.reason) {
      case 'deadline-exceeded':
        return { kind: 'safe-state', reason: 'deadline-exceeded' };
      case 'not-calibrated':
        return { kind: 'safe-state', reason: 'not-ready' };
      case 'observation-unavailable':
        return { kind: 'safe-state', reason: 'stale-observation' };
      case 'no-candidates':
        return { kind: 'safe-state', reason: 'no-candidates' };
      default:
        return { kind: 'safe-state', reason: 'not-ready' };
    }

  } catch (error) {
    throw new Error(`[advisoryDecisionOfCertified] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Handles the gnosis advisory Allocation Core workflow.
 */
export function advisoryAllocationCore(
  plan: CertifiedCalibrationPlan,
  points: readonly CertifiedCalibrationPointLedger[],
  modelSolved: boolean,
  ingestLocked: boolean,
  frame: AdvisoryObservationFrame,
  candidateIds: readonly number[],
  risk: RiskEnvelope,
  budget: AdvisoryBudget
): AdvisoryAllocationDecision {
  try {
    if (!withinAdvisoryBudget(budget)) {
      return { kind: 'safe-state', reason: 'deadline-exceeded' };
    }

    if (!marketReady(plan, points, modelSolved, ingestLocked)) {
      return { kind: 'safe-state', reason: 'not-ready' };
    }

    if (!riskWithinLimits(risk)) {
      return { kind: 'safe-state', reason: 'risk-limit-reached' };
    }

    if (!observationFresh(frame)) {
      return { kind: 'safe-state', reason: 'stale-observation' };
    }

    const decayedFrame = decayedObservationFrame(frame);
    if (!confidenceFloorSatisfied(decayedFrame)) {
      return { kind: 'safe-state', reason: 'stale-observation' };
    }

    if (!stabilityWitness(frame)) {
      return { kind: 'safe-state', reason: 'adverse-selection' };
    }

    return advisoryDecisionOfCertified(
      certifiedAllocationStep(
        plan,
        points,
        modelSolved,
        decayedFrame,
        candidateIds,
        advisoryLoopBudget(budget)
      )
    );

  } catch (error) {
    throw new Error(`[advisoryAllocationCore] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Handles the gnosis advisory Allocation Step workflow.
 */
export function advisoryAllocationStep(
  plan: CertifiedCalibrationPlan,
  points: readonly CertifiedCalibrationPointLedger[],
  modelSolved: boolean,
  ingestLocked: boolean,
  frame: AdvisoryObservationFrame,
  candidateIds: readonly number[],
  risk: RiskEnvelope,
  budget: AdvisoryBudget
): AdvisoryAllocationDecision {
  try {
    if (commitPointReached(budget) && !marketReady(plan, points, modelSolved, ingestLocked)) {
      return { kind: 'safe-state', reason: 'deadline-exceeded' };
    }

    return advisoryAllocationCore(
      plan,
      points,
      modelSolved,
      ingestLocked,
      frame,
      candidateIds,
      risk,
      budget
    );

  } catch (error) {
    throw new Error(`[advisoryAllocationStep] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Handles the gnosis derive Advisory Allocation Kernel workflow.
 */
export function deriveAdvisoryAllocationKernel(
  input: AdvisoryAllocationKernelInput
): AdvisoryAllocationKernelBundle {
  const normalizedPlan = {
    pointIds: input.plan.pointIds.map((pointId) => pointId),
    samplesPerPoint: clampNonNegativeInt(input.plan.samplesPerPoint),
    maxAttemptsPerPoint: clampNonNegativeInt(input.plan.maxAttemptsPerPoint),
  } satisfies CertifiedCalibrationPlan;
  const normalizedPoints = input.points.map((point) => ({
    pointId: point.pointId,
    accepted: clampNonNegativeInt(point.accepted),
    rejected: clampNonNegativeInt(point.rejected),
    observed: clampNonNegativeInt(point.observed),
  }));
  const normalizedFrame = normalizeFrame(input.frame);
  const normalizedRisk = normalizeRisk(input.risk);
  const normalizedBudget = normalizeBudget(input.budget);
  const normalizedCandidates = normalizeCandidates(input.candidateIds);
  const certifiedBridgeDecision = certifiedAllocationStep(
    normalizedPlan,
    normalizedPoints,
    input.modelSolved,
    decayedObservationFrame(normalizedFrame),
    normalizedCandidates,
    advisoryLoopBudget(normalizedBudget)
  );

  return {
    generatedAt: input.generatedAt ?? nowIso(),
    source: input.source?.trim() || 'advisory-allocation-kernel',
    theoremIds: ADVISORY_ALLOCATION_KERNEL_THEOREM_IDS,
    decision: advisoryAllocationStep(
      normalizedPlan,
      normalizedPoints,
      input.modelSolved,
      input.ingestLocked,
      normalizedFrame,
      normalizedCandidates,
      normalizedRisk,
      normalizedBudget
    ),
    aggregate: {
      priceMove: priceMove(normalizedFrame),
      confidenceDecay: confidenceDecay(normalizedFrame),
      decayedConfidenceScore: decayedConfidenceScore(normalizedFrame),
      observationFresh: observationFresh(normalizedFrame),
      stabilityWitness: stabilityWitness(normalizedFrame),
      riskWithinLimits: riskWithinLimits(normalizedRisk),
      marketReady: marketReady(
        normalizedPlan,
        normalizedPoints,
        input.modelSolved,
        input.ingestLocked
      ),
      withinBudget: withinAdvisoryBudget(normalizedBudget),
      commitPointReached: commitPointReached(normalizedBudget),
      confidenceFloorSatisfied: confidenceFloorSatisfied(
        decayedObservationFrame(normalizedFrame)
      ),
      certifiedBridgeDecision,
      inheritedCertifiedTheoremIds: CERTIFIED_ALLOCATION_KERNEL_THEOREM_IDS,
    },
  };
}

/**
 * Creates the Synthetic Visible Observation.
 */
export function createSyntheticVisibleObservation(
  confidenceScore: number,
  confidenceFloor: number
): CertifiedObservationFrame {
  return {
    gazePoint: SYNTHETIC_VISIBLE_POINT,
    zone: 'center',
    confidenceScore: clampNonNegativeInt(confidenceScore),
    confidenceFloor: clampNonNegativeInt(confidenceFloor),
  };
}
