// @a0n/.ignored_gnosis/src/apm.ts
export type TraceEnv = Record<string, unknown>;
export type TraceAttributes = Record<string, string | number | boolean>;

export interface Span {
  setAttribute(key: string, value: string | number | boolean): void;
  setAttributes(attributes: TraceAttributes): void;
  end(): void;
}

export type TraceProviderLike = {
  register?: () => void;
  shutdown?: () => Promise<void>;
  addSpanProcessor?: (spanProcessor: unknown) => void;
};

interface TraceExporterTarget {
  readonly endpoint: string;
  readonly headers?: Record<string, string>;
}

interface TelemetryModule {
  createNodeTracerProvider(options?: {
    endpoint?: string;
    headers?: Record<string, string>;
    debug?: boolean;
    resourceAttributes?: Record<string, string>;
  }): TraceProviderLike;
  getTracer(name: string): unknown;
  registerTracerProvider(provider: TraceProviderLike): void;
  resolveTraceExporterTarget(
    env: TraceEnv | undefined,
    fallback?: string
  ): TraceExporterTarget;
  withSpan<T>(
    tracer: unknown,
    name: string,
    fn: (span: Span) => Promise<T> | T,
    attributes?: TraceAttributes
  ): Promise<T>;
}

type RuntimeRequire = (specifier: string) => unknown;
type RuntimeCreateRequire = (baseUrl: string) => RuntimeRequire;
type NodeProcessWithBuiltinModule = NodeJS.Process & {
  getBuiltinModule?: (id: string) => unknown;
};
type NodeModuleBuiltin = {
  createRequire?: RuntimeCreateRequire;
};

function resolveRuntimeRequire(): RuntimeRequire | null {
  if (typeof require !== 'undefined') {
    return require as RuntimeRequire;
  }

  const runtimeProcess =
    typeof process !== 'undefined'
      ? (process as NodeProcessWithBuiltinModule)
      : null;
  const nodeModule = runtimeProcess?.getBuiltinModule?.(
    'module'
  ) as NodeModuleBuiltin | null | undefined;
  const createRequire =
    typeof nodeModule?.createRequire === 'function'
      ? nodeModule.createRequire
      : null;

  if (!createRequire || typeof import.meta === 'undefined' || !import.meta.url) {
    return null;
  }

  return createRequire(import.meta.url);
}

const getRequire = () => {
  try {
    return resolveRuntimeRequire();
  } catch {
    return null;
  }
};
const customRequire = getRequire();

const NOOP_SPAN: Span = {
  setAttribute() {},
  setAttributes() {},
  end() {},
};

let telemetryModule: TelemetryModule | null = null;
let telemetryResolutionAttempted = false;

function resolveTelemetryModule(): TelemetryModule | null {
  if (telemetryResolutionAttempted) {
    return telemetryModule;
  }
  telemetryResolutionAttempted = true;

  try {
    if (customRequire) {
      const mod = customRequire('@a0n/telemetry') as TelemetryModule;
      telemetryModule = typeof mod?.getTracer === 'function' ? mod : null;
    } else {
      telemetryModule = null;
    }
  } catch {
    telemetryModule = null;
  }

  return telemetryModule;
}

export const tracer = resolveTelemetryModule()?.getTracer('gnosis-cli') ?? null;

let provider: TraceProviderLike | null = null;
let tracingUnavailable = false;

function readEnvString(env: TraceEnv | undefined, key: string): string | null {
  try {
    const value = env?.[key];
    if (typeof value !== 'string') {
      return null;
    }

    const normalized = value.trim();
    return normalized.length > 0 ? normalized : null;

  } catch (error) {
    throw new Error(`[readEnvString] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function readFirstEnvString(
  env: TraceEnv | undefined,
  keys: readonly string[]
): string | null {
  try {
    for (const key of keys) {
      const value = readEnvString(env, key);
      if (value) {
        return value;
      }
    }

    return null;

  } catch (error) {
    throw new Error(`[readFirstEnvString] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function parseBooleanFlag(value: string | null): boolean | null {
  try {
    if (!value) {
      return null;
    }

    const normalized = value.toLowerCase();
    if (
      normalized === '1' ||
      normalized === 'true' ||
      normalized === 'yes' ||
      normalized === 'on'
    ) {
      return true;
    }
    if (
      normalized === '0' ||
      normalized === 'false' ||
      normalized === 'no' ||
      normalized === 'off'
    ) {
      return false;
    }

    return null;

  } catch (error) {
    throw new Error(`[parseBooleanFlag] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function hasAnyEnvValue(env: TraceEnv | undefined, keys: readonly string[]): boolean {
  return keys.some((key) => readEnvString(env, key) !== null);
}

function resolveResourceAttributes(env: TraceEnv | undefined): Record<string, string> {
  try {
    const attributes: Record<string, string> = {
      'service.name':
        readFirstEnvString(env, ['GNOSIS_SERVICE_NAME', 'OTEL_SERVICE_NAME', 'DD_SERVICE']) ??
        'gnosis',
    };

    const serviceVersion = readFirstEnvString(env, [
      'GNOSIS_SERVICE_VERSION',
      'DD_VERSION',
    ]);
    if (serviceVersion) {
      attributes['service.version'] = serviceVersion;
    }

    const deploymentEnvironment = readFirstEnvString(env, [
      'GNOSIS_ENV',
      'DD_ENV',
      'NODE_ENV',
    ]);
    if (deploymentEnvironment) {
      attributes['deployment.environment'] = deploymentEnvironment;
    }

    return attributes;

  } catch (error) {
    throw new Error(`[resolveResourceAttributes] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Returns whether is Gnosis Tracing Enabled is true.
 */
export function isGnosisTracingEnabled(env: TraceEnv | undefined): boolean {
  try {
    const explicit = parseBooleanFlag(
      readEnvString(env, 'GNOSIS_TRACING_ENABLED')
    );
    if (explicit !== null) {
      return explicit;
    }

    const datadogFlag = parseBooleanFlag(
      readFirstEnvString(env, ['DATADOG_APM_ENABLED', 'DD_APM_ENABLED'])
    );
    if (datadogFlag === true) {
      return true;
    }

    return (
      hasAnyEnvValue(env, [
        'OTEL_EXPORTER_OTLP_ENDPOINT',
        'OTEL_EXPORTER_OTLP_TRACES_ENDPOINT',
        'DATADOG_OTLP_ENDPOINT',
        'DATADOG_OTLP_TRACES_ENDPOINT',
        'DD_OTLP_ENDPOINT',
        'DD_OTLP_TRACES_ENDPOINT',
        'DD_API_KEY',
        'DATADOG_API_KEY',
      ])
    );

  } catch (error) {
    throw new Error(`[isGnosisTracingEnabled] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Handles the gnosis init Gnosis Tracing From Env workflow.
 */
export function initGnosisTracingFromEnv(
  env: TraceEnv | undefined = process.env
): void {
  if (provider || tracingUnavailable || !isGnosisTracingEnabled(env)) {
    return;
  }

  const telemetry = resolveTelemetryModule();
  if (!telemetry) {
    tracingUnavailable = true;
    return;
  }

  try {
    const exporterTarget = telemetry.resolveTraceExporterTarget(env);
    provider = telemetry.createNodeTracerProvider({
      endpoint: exporterTarget.endpoint,
      ...(exporterTarget.headers ? { headers: exporterTarget.headers } : {}),
      debug: readEnvString(env, 'DEBUG_TELEMETRY') === 'true',
      resourceAttributes: resolveResourceAttributes(env),
    });
    telemetry.registerTracerProvider(provider);
  } catch (error: unknown) {
    tracingUnavailable = true;
    console.error('[gnosis] Failed to initialize tracing:', error);
  }
}

export async function withGnosisSpan<T>(
  name: string,
  fn: (span: Span) => Promise<T> | T,
  attributes?: TraceAttributes
): Promise<T> {
  try {
    const telemetry = resolveTelemetryModule();
    if (!telemetry) {
      return Promise.resolve(fn(NOOP_SPAN));
    }

    return telemetry.withSpan(tracer, name, fn, attributes);

  } catch (error) {
    throw new Error(`[withGnosisSpan] ${error instanceof Error ? error.message : String(error)}`);
  }
}
