// @a0n/.ignored_gnosis/src/avatar.ts
export const GNOSIS_AVATAR_HUMANOID_BONES = [
  'hips',
  'spine',
  'chest',
  'upperChest',
  'neck',
  'head',
  'jaw',
  'leftEye',
  'rightEye',
  'leftShoulder',
  'leftUpperArm',
  'leftLowerArm',
  'leftHand',
  'rightShoulder',
  'rightUpperArm',
  'rightLowerArm',
  'rightHand',
  'leftUpperLeg',
  'leftLowerLeg',
  'leftFoot',
  'rightUpperLeg',
  'rightLowerLeg',
  'rightFoot',
] as const;

export type GnosisAvatarHumanoidBone =
  (typeof GNOSIS_AVATAR_HUMANOID_BONES)[number];

export type GnosisAvatarClipName =
  | 'idle'
  | 'listen'
  | 'think'
  | 'speak'
  | 'settle';

export interface GnosisAvatarVector3 {
  x: number;
  y: number;
  z: number;
}

export interface GnosisAvatarQuaternion {
  x: number;
  y: number;
  z: number;
  w: number;
}

export interface GnosisAvatarMorphSource {
  nodeIndex: number;
  meshName?: string;
  targetNames: string[];
}

export interface GnosisVrmMorphTargetBind {
  node: number;
  index: number;
  weight: number;
}

export interface GnosisVrmExpressionPreset {
  morphTargetBinds?: GnosisVrmMorphTargetBind[];
  isBinary?: boolean;
}

export interface GnosisVrmHumanBoneDescriptor {
  node: number;
}

export interface GnosisVrmRangeMap {
  inputMaxValue?: number;
  outputScale?: number;
}

export interface GnosisVrmExtensionLike {
  meta?: {
    name?: string;
    version?: string;
    authors?: string[];
  };
  humanoid?: {
    humanBones?: Partial<
      Record<GnosisAvatarHumanoidBone, GnosisVrmHumanBoneDescriptor>
    >;
  };
  lookAt?: {
    rangeMapHorizontalInner?: GnosisVrmRangeMap;
    rangeMapHorizontalOuter?: GnosisVrmRangeMap;
    rangeMapVerticalDown?: GnosisVrmRangeMap;
    rangeMapVerticalUp?: GnosisVrmRangeMap;
  };
  expressions?: {
    preset?: Record<string, GnosisVrmExpressionPreset>;
    custom?: Record<string, GnosisVrmExpressionPreset>;
  };
}

export interface GnosisAvatarCompilationInput {
  nodeNames: string[];
  morphSources: GnosisAvatarMorphSource[];
  vrm?: GnosisVrmExtensionLike | null;
  clipNames?: string[];
}

export interface GnosisAvatarBoneSlot {
  slot: number;
  semantic: GnosisAvatarHumanoidBone;
  sourceNodeIndex: number;
  name: string;
}

export interface GnosisAvatarMorphSlot {
  slot: number;
  sourceNodeIndex: number;
  targetIndex: number;
  meshName?: string;
  name: string;
  normalizedName: string;
  role: 'expression' | 'viseme' | 'blink' | 'unknown';
}

export interface GnosisAvatarExpressionProgram {
  name: string;
  binary: boolean;
  morphBinds: Array<{
    slot: number;
    weight: number;
  }>;
}

export interface GnosisAvatarLookAtProfile {
  yawLimitDeg: number;
  pitchUpLimitDeg: number;
  pitchDownLimitDeg: number;
  headWeight: number;
  eyeWeight: number;
}

export interface GnosisAvatarClipProgram {
  name: GnosisAvatarClipName;
  cadenceHz: number;
  intensity: number;
}

export interface GnosisAvatarRuntimeManifest {
  source: {
    modelName?: string;
    version?: string;
    authors: string[];
  };
  bones: GnosisAvatarBoneSlot[];
  morphs: GnosisAvatarMorphSlot[];
  expressions: GnosisAvatarExpressionProgram[];
  visemeAliases: Record<string, number[]>;
  lookAt: GnosisAvatarLookAtProfile;
  clips: GnosisAvatarClipProgram[];
  diagnostics: string[];
}

export interface GnosisAvatarSignalFrame {
  timestampMs: number;
  gazeTarget?: GnosisAvatarVector3;
  blink?: number;
  lookAtWeight?: number;
  expressionWeights?: Record<string, number>;
  visemes?: Record<string, number>;
  emotion?: {
    valence?: number;
    arousal?: number;
    dominance?: number;
    stress?: number;
    focus?: number;
    dominantEmotion?: string;
  };
  posture?: 'neutral' | 'open' | 'contracted' | 'engaged' | 'fatigued';
  gestureIntent?: 'idle' | 'listen' | 'speak' | 'think' | 'settle' | 'emphasis';
  speechState?: 'idle' | 'listening' | 'thinking' | 'speaking';
  confidence?: number;
}

export interface GnosisAvatarBoneDelta {
  slot: number;
  rotation: GnosisAvatarQuaternion;
  positionOffset?: GnosisAvatarVector3;
}

export interface GnosisAvatarMorphDelta {
  slot: number;
  weight: number;
}

export interface GnosisAvatarAuraState {
  energy: number;
  turbulence: number;
  focus: number;
  tint: [number, number, number];
}

export interface GnosisAvatarPoseResult {
  activeClip: GnosisAvatarClipName;
  boneDeltas: GnosisAvatarBoneDelta[];
  morphDeltas: GnosisAvatarMorphDelta[];
  aura: GnosisAvatarAuraState;
}

const QUATERNION_IDENTITY: GnosisAvatarQuaternion = {
  x: 0,
  y: 0,
  z: 0,
  w: 1,
};

const BONE_ALIASES: Record<GnosisAvatarHumanoidBone, string[]> = {
  hips: ['hips', 'pelvis'],
  spine: ['spine'],
  chest: ['chest'],
  upperChest: ['upperchest', 'upper_chest'],
  neck: ['neck'],
  head: ['head'],
  jaw: ['jaw', 'mandible'],
  leftEye: ['lefteye', 'eye_l'],
  rightEye: ['righteye', 'eye_r'],
  leftShoulder: ['leftshoulder', 'shoulder_l'],
  leftUpperArm: ['leftupperarm', 'upperarm_l', 'leftarm'],
  leftLowerArm: ['leftlowerarm', 'lowerarm_l', 'leftforearm'],
  leftHand: ['lefthand', 'hand_l'],
  rightShoulder: ['rightshoulder', 'shoulder_r'],
  rightUpperArm: ['rightupperarm', 'upperarm_r', 'rightarm'],
  rightLowerArm: ['rightlowerarm', 'lowerarm_r', 'rightforearm'],
  rightHand: ['righthand', 'hand_r'],
  leftUpperLeg: ['leftupperleg', 'upleg_l', 'leftthigh'],
  leftLowerLeg: ['leftlowerleg', 'leg_l', 'leftshin'],
  leftFoot: ['leftfoot', 'foot_l'],
  rightUpperLeg: ['rightupperleg', 'upleg_r', 'rightthigh'],
  rightLowerLeg: ['rightlowerleg', 'leg_r', 'rightshin'],
  rightFoot: ['rightfoot', 'foot_r'],
};

const DEFAULT_CLIPS: GnosisAvatarClipProgram[] = [
  { name: 'idle', cadenceHz: 0.35, intensity: 0.3 },
  { name: 'listen', cadenceHz: 0.5, intensity: 0.5 },
  { name: 'think', cadenceHz: 0.28, intensity: 0.45 },
  { name: 'speak', cadenceHz: 0.9, intensity: 0.75 },
  { name: 'settle', cadenceHz: 0.2, intensity: 0.2 },
];

const EMOTION_EXPRESSION_ALIASES: Record<string, string[]> = {
  happy: ['happy', 'joy', 'smile'],
  angry: ['angry', 'anger'],
  sad: ['sad', 'sorrow'],
  surprised: ['surprised', 'surprise'],
  fearful: ['fear', 'fearful'],
  blink: ['blink', 'blinkleft', 'blinkright', 'eyesclosed'],
};

const VISEME_ALIASES: Record<string, string[]> = {
  aa: ['aa', 'a'],
  ee: ['ee', 'e'],
  ih: ['ih', 'i'],
  oh: ['oh', 'o'],
  ou: ['ou', 'u'],
};

function matchesAlias(name: string, alias: string): boolean {
  try {
    if (name === alias) {
      return true;
    }
    if (alias.length <= 1) {
      return false;
    }
    return name.startsWith(alias) || name.endsWith(alias);

  } catch (error) {
    throw new Error(`[matchesAlias] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function clampUnit(value: number): number {
  try {
    if (Number.isNaN(value)) {
      return 0;
    }
    return Math.min(1, Math.max(0, value));

  } catch (error) {
    throw new Error(`[clampUnit] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function clamp(value: number, min: number, max: number): number {
  return Math.min(max, Math.max(min, value));
}

function normalizeName(value: string): string {
  return value.toLowerCase().replace(/[^a-z0-9]+/g, '');
}

function degreesToRadians(value: number): number {
  return (value * Math.PI) / 180;
}

function axisAngleToQuaternion(
  axis: 'x' | 'y' | 'z',
  radians: number
): GnosisAvatarQuaternion {
  try {
    const half = radians / 2;
    const sine = Math.sin(half);
    const cosine = Math.cos(half);
    if (axis === 'x') {
      return { x: sine, y: 0, z: 0, w: cosine };
    }
    if (axis === 'y') {
      return { x: 0, y: sine, z: 0, w: cosine };
    }
    return { x: 0, y: 0, z: sine, w: cosine };

  } catch (error) {
    throw new Error(`[axisAngleToQuaternion] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function multiplyQuaternions(
  left: GnosisAvatarQuaternion,
  right: GnosisAvatarQuaternion
): GnosisAvatarQuaternion {
  return {
    w:
      left.w * right.w -
      left.x * right.x -
      left.y * right.y -
      left.z * right.z,
    x:
      left.w * right.x +
      left.x * right.w +
      left.y * right.z -
      left.z * right.y,
    y:
      left.w * right.y -
      left.x * right.z +
      left.y * right.w +
      left.z * right.x,
    z:
      left.w * right.z +
      left.x * right.y -
      left.y * right.x +
      left.z * right.w,
  };
}

function detectMorphRole(name: string): GnosisAvatarMorphSlot['role'] {
  try {
    const blinkAliases = EMOTION_EXPRESSION_ALIASES.blink ?? [];
    if (blinkAliases.some((alias) => matchesAlias(name, alias))) {
      return 'blink';
    }
    if (
      Object.values(VISEME_ALIASES).some((aliases) =>
        aliases.some((alias) => matchesAlias(name, alias))
      )
    ) {
      return 'viseme';
    }
    if (
      Object.values(EMOTION_EXPRESSION_ALIASES).some((aliases) =>
        aliases.some((alias) => matchesAlias(name, alias))
      )
    ) {
      return 'expression';
    }
    return 'unknown';

  } catch (error) {
    throw new Error(`[detectMorphRole] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function getMorphSlotBySource(
  morphs: GnosisAvatarMorphSlot[],
  nodeIndex: number,
  targetIndex: number
): GnosisAvatarMorphSlot | undefined {
  return morphs.find(
    (morph) =>
      morph.sourceNodeIndex === nodeIndex && morph.targetIndex === targetIndex
  );
}

function getBoneSlot(
  manifest: GnosisAvatarRuntimeManifest,
  semantic: GnosisAvatarHumanoidBone
): GnosisAvatarBoneSlot | undefined {
  return manifest.bones.find((bone) => bone.semantic === semantic);
}

function addMergedBoneDelta(
  target: Map<number, GnosisAvatarQuaternion>,
  slot: number,
  rotation: GnosisAvatarQuaternion
): void {
  const existing = target.get(slot) ?? QUATERNION_IDENTITY;
  target.set(slot, multiplyQuaternions(existing, rotation));
}

function inferExpressionPrograms(
  morphs: GnosisAvatarMorphSlot[]
): GnosisAvatarExpressionProgram[] {
  return Object.entries(EMOTION_EXPRESSION_ALIASES)
    .flatMap(([name, aliases]) => {
      const matched = morphs.filter((morph) =>
        aliases.some((alias) => matchesAlias(morph.normalizedName, alias))
      );
      if (matched.length === 0) {
        return [];
      }
      return [
        {
          name,
          binary: name === 'blink',
          morphBinds: matched.map((morph) => ({ slot: morph.slot, weight: 1 })),
        },
      ];
    })
    .filter(
      (program, index, programs) =>
        programs.findIndex((candidate) => candidate.name === program.name) ===
        index
    );
}

function buildVisemeAliases(
  morphs: GnosisAvatarMorphSlot[]
): Record<string, number[]> {
  return Object.fromEntries(
    Object.entries(VISEME_ALIASES).map(([viseme, aliases]) => [
      viseme,
      morphs
        .filter((morph) =>
          aliases.some((alias) => matchesAlias(morph.normalizedName, alias))
        )
        .map((morph) => morph.slot),
    ])
  );
}

function chooseClip(frame: GnosisAvatarSignalFrame): GnosisAvatarClipName {
  try {
    if (frame.speechState === 'speaking' || frame.gestureIntent === 'speak') {
      return 'speak';
    }
    if (frame.speechState === 'thinking' || frame.gestureIntent === 'think') {
      return 'think';
    }
    if (frame.speechState === 'listening' || frame.gestureIntent === 'listen') {
      return 'listen';
    }
    if (frame.gestureIntent === 'settle') {
      return 'settle';
    }
    return 'idle';

  } catch (error) {
    throw new Error(`[chooseClip] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function getClipIntensity(
  manifest: GnosisAvatarRuntimeManifest,
  clipName: GnosisAvatarClipName
): number {
  return manifest.clips.find((clip) => clip.name === clipName)?.intensity ?? 0.4;
}

function deriveEmotionExpressionWeights(
  frame: GnosisAvatarSignalFrame
): Record<string, number> {
  try {
    const emotion = frame.emotion;
    if (!emotion) {
      return {};
    }

    const dominant = normalizeName(emotion.dominantEmotion ?? '');
    if (dominant) {
      return {
        [dominant]: clampUnit(frame.confidence ?? 0.5),
      };
    }

    if ((emotion.valence ?? 0) > 0.25) {
      return { happy: clampUnit(emotion.valence ?? 0) };
    }
    if ((emotion.valence ?? 0) < -0.25 && (emotion.arousal ?? 0) > 0.45) {
      return { angry: clampUnit(Math.abs(emotion.valence ?? 0)) };
    }
    if ((emotion.valence ?? 0) < -0.25) {
      return { sad: clampUnit(Math.abs(emotion.valence ?? 0)) };
    }

    return {};

  } catch (error) {
    throw new Error(`[deriveEmotionExpressionWeights] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function tintFromEmotion(
  emotion: GnosisAvatarSignalFrame['emotion']
): [number, number, number] {
  try {
    const dominant = normalizeName(emotion?.dominantEmotion ?? '');
    if (dominant.includes('angry')) {
      return [1, 0.28, 0.22];
    }
    if (dominant.includes('sad')) {
      return [0.35, 0.55, 1];
    }
    if (dominant.includes('happy')) {
      return [1, 0.82, 0.3];
    }
    return [0.58, 0.88, 1];

  } catch (error) {
    throw new Error(`[tintFromEmotion] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Handles the gnosis compile Vrm Avatar Runtime Manifest workflow.
 */
export function compileVrmAvatarRuntimeManifest(
  input: GnosisAvatarCompilationInput
): GnosisAvatarRuntimeManifest {
  try {
    const diagnostics: string[] = [];
    const bones: GnosisAvatarBoneSlot[] = [];
    const normalizedNodes = input.nodeNames.map((name) => normalizeName(name));
    const vrmHumanBones = input.vrm?.humanoid?.humanBones;

    for (const semantic of GNOSIS_AVATAR_HUMANOID_BONES) {
      const humanBoneNode = vrmHumanBones?.[semantic]?.node;
      const sourceNodeIndex =
        typeof humanBoneNode === 'number'
          ? humanBoneNode
          : normalizedNodes.findIndex((nodeName) =>
              BONE_ALIASES[semantic].some((alias) => nodeName.includes(alias))
            );

      if (sourceNodeIndex < 0) {
        continue;
      }

      bones.push({
        slot: bones.length,
        semantic,
        sourceNodeIndex,
        name:
          input.nodeNames[sourceNodeIndex] ??
          `${semantic}-${String(sourceNodeIndex)}`,
      });
    }

    if (!bones.some((bone) => bone.semantic === 'head')) {
      diagnostics.push('Missing head bone; look-at will degrade to no-op.');
    }

    const morphs: GnosisAvatarMorphSlot[] = input.morphSources.flatMap(
      (source) =>
        source.targetNames.map((targetName, targetIndex) => {
          const normalizedName = normalizeName(targetName);
          return {
            slot: 0,
            sourceNodeIndex: source.nodeIndex,
            targetIndex,
            name: targetName,
            normalizedName,
            role: detectMorphRole(normalizedName),
            ...(source.meshName ? { meshName: source.meshName } : {}),
          };
        })
    );

    morphs.forEach((morph, slot) => {
      morph.slot = slot;
    });

    const vrmExpressions = [
      ...(input.vrm?.expressions?.preset
        ? Object.entries(input.vrm.expressions.preset)
        : []),
      ...(input.vrm?.expressions?.custom
        ? Object.entries(input.vrm.expressions.custom)
        : []),
    ].flatMap(([name, expression]) => {
      const morphBinds = expression.morphTargetBinds
        ?.map((bind) => {
          const morphSlot = getMorphSlotBySource(morphs, bind.node, bind.index);
          return morphSlot
            ? { slot: morphSlot.slot, weight: clampUnit(bind.weight / 100) }
            : null;
        })
        .filter((bind): bind is { slot: number; weight: number } => bind !== null);

      if (!morphBinds || morphBinds.length === 0) {
        return [];
      }

      return [
        {
          name: normalizeName(name),
          binary: expression.isBinary ?? false,
          morphBinds,
        },
      ];
    });

    const inferredExpressions = inferExpressionPrograms(morphs);
    const expressions = [...vrmExpressions];
    for (const inferred of inferredExpressions) {
      if (!expressions.some((expression) => expression.name === inferred.name)) {
        expressions.push(inferred);
      }
    }

    const lookAt: GnosisAvatarLookAtProfile = {
      yawLimitDeg:
        (input.vrm?.lookAt?.rangeMapHorizontalOuter?.inputMaxValue ?? 45) *
        (input.vrm?.lookAt?.rangeMapHorizontalOuter?.outputScale ?? 1),
      pitchUpLimitDeg:
        (input.vrm?.lookAt?.rangeMapVerticalUp?.inputMaxValue ?? 20) *
        (input.vrm?.lookAt?.rangeMapVerticalUp?.outputScale ?? 1),
      pitchDownLimitDeg:
        (input.vrm?.lookAt?.rangeMapVerticalDown?.inputMaxValue ?? 25) *
        (input.vrm?.lookAt?.rangeMapVerticalDown?.outputScale ?? 1),
      headWeight: 0.55,
      eyeWeight: 0.45,
    };

    return {
      source: {
        ...(input.vrm?.meta?.name ? { modelName: input.vrm.meta.name } : {}),
        ...(input.vrm?.meta?.version ? { version: input.vrm.meta.version } : {}),
        authors: input.vrm?.meta?.authors ?? [],
      },
      bones,
      morphs,
      expressions,
      visemeAliases: buildVisemeAliases(morphs),
      lookAt,
      clips:
        (input.clipNames ?? []).length > 0
          ? DEFAULT_CLIPS.filter((clip) =>
              (input.clipNames ?? []).includes(clip.name)
            )
          : DEFAULT_CLIPS,
      diagnostics,
    };

  } catch (error) {
    throw new Error(`[compileVrmAvatarRuntimeManifest] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Handles the gnosis solve Look At Bone Deltas workflow.
 */
export function solveLookAtBoneDeltas(
  manifest: GnosisAvatarRuntimeManifest,
  target: GnosisAvatarVector3,
  weight = 1
): GnosisAvatarBoneDelta[] {
  try {
    const vectorLength = Math.sqrt(
      target.x * target.x + target.y * target.y + target.z * target.z
    );
    if (vectorLength === 0) {
      return [];
    }

    const yaw = clamp(
      (Math.atan2(target.x, target.z) * 180) / Math.PI,
      -manifest.lookAt.yawLimitDeg,
      manifest.lookAt.yawLimitDeg
    );
    const pitch = clamp(
      (Math.atan2(target.y, Math.sqrt(target.x * target.x + target.z * target.z)) *
        180) /
        Math.PI,
      -manifest.lookAt.pitchDownLimitDeg,
      manifest.lookAt.pitchUpLimitDeg
    );

    const merged = new Map<number, GnosisAvatarQuaternion>();
    const scaledWeight = clampUnit(weight);
    const head = getBoneSlot(manifest, 'head');
    const leftEye = getBoneSlot(manifest, 'leftEye');
    const rightEye = getBoneSlot(manifest, 'rightEye');

    const headYaw = axisAngleToQuaternion(
      'y',
      degreesToRadians(yaw * manifest.lookAt.headWeight * scaledWeight)
    );
    const headPitch = axisAngleToQuaternion(
      'x',
      degreesToRadians(-pitch * manifest.lookAt.headWeight * scaledWeight)
    );
    const eyeYaw = axisAngleToQuaternion(
      'y',
      degreesToRadians(yaw * manifest.lookAt.eyeWeight * scaledWeight)
    );
    const eyePitch = axisAngleToQuaternion(
      'x',
      degreesToRadians(-pitch * manifest.lookAt.eyeWeight * scaledWeight)
    );

    if (head) {
      addMergedBoneDelta(merged, head.slot, headYaw);
      addMergedBoneDelta(merged, head.slot, headPitch);
    }
    if (leftEye) {
      addMergedBoneDelta(merged, leftEye.slot, eyeYaw);
      addMergedBoneDelta(merged, leftEye.slot, eyePitch);
    }
    if (rightEye) {
      addMergedBoneDelta(merged, rightEye.slot, eyeYaw);
      addMergedBoneDelta(merged, rightEye.slot, eyePitch);
    }

    return Array.from(merged.entries()).map(([slot, rotation]) => ({
      slot,
      rotation,
    }));

  } catch (error) {
    throw new Error(`[solveLookAtBoneDeltas] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Handles the gnosis sample Procedural Avatar Clip workflow.
 */
export function sampleProceduralAvatarClip(
  manifest: GnosisAvatarRuntimeManifest,
  clipName: GnosisAvatarClipName,
  timestampMs: number,
  intensity = getClipIntensity(manifest, clipName)
): GnosisAvatarBoneDelta[] {
  try {
    const phase = (timestampMs / 1_000) * (getClipIntensity(manifest, clipName) + 0.2);
    const head = getBoneSlot(manifest, 'head');
    const neck = getBoneSlot(manifest, 'neck');
    const chest = getBoneSlot(manifest, 'chest');
    const jaw = getBoneSlot(manifest, 'jaw');
    const deltas: GnosisAvatarBoneDelta[] = [];

    if (head) {
      const headRoll = axisAngleToQuaternion(
        'z',
        Math.sin(phase * Math.PI) * intensity * 0.08
      );
      const headNod = axisAngleToQuaternion(
        'x',
        Math.cos(phase * Math.PI * 0.6) * intensity * 0.05
      );
      deltas.push({
        slot: head.slot,
        rotation: multiplyQuaternions(headRoll, headNod),
      });
    }

    if (neck && (clipName === 'listen' || clipName === 'think')) {
      deltas.push({
        slot: neck.slot,
        rotation: axisAngleToQuaternion(
          'x',
          (clipName === 'listen' ? -1 : 1) * intensity * 0.04
        ),
      });
    }

    if (chest) {
      deltas.push({
        slot: chest.slot,
        rotation: axisAngleToQuaternion(
          'y',
          Math.sin(phase * Math.PI * 0.25) * intensity * 0.03
        ),
        positionOffset: {
          x: 0,
          y: Math.sin(phase * Math.PI * 0.4) * intensity * 0.008,
          z: 0,
        },
      });
    }

    if (jaw && clipName === 'speak') {
      deltas.push({
        slot: jaw.slot,
        rotation: axisAngleToQuaternion(
          'x',
          (-0.08 - Math.abs(Math.sin(phase * Math.PI * 3)) * 0.12) * intensity
        ),
      });
    }

    return deltas;

  } catch (error) {
    throw new Error(`[sampleProceduralAvatarClip] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Handles the gnosis mix Avatar Morph Deltas workflow.
 */
export function mixAvatarMorphDeltas(
  manifest: GnosisAvatarRuntimeManifest,
  frame: GnosisAvatarSignalFrame
): GnosisAvatarMorphDelta[] {
  try {
    const weights = new Map<number, number>();
    const applySlots = (slots: number[], weight: number): void => {
      for (const slot of slots) {
        const nextWeight = Math.max(weights.get(slot) ?? 0, clampUnit(weight));
        weights.set(slot, nextWeight);
      }
    };
    const applyExpression = (name: string, weight: number): void => {
      const expression = manifest.expressions.find(
        (candidate) => candidate.name === normalizeName(name)
      );
      if (!expression) {
        return;
      }

      for (const morphBind of expression.morphBinds) {
        const nextWeight = expression.binary
          ? weight > 0.5
            ? morphBind.weight
            : 0
          : clampUnit(weight * morphBind.weight);
        weights.set(
          morphBind.slot,
          Math.max(weights.get(morphBind.slot) ?? 0, nextWeight)
        );
      }
    };

    const explicitExpressions = frame.expressionWeights ?? {};
    for (const [name, weight] of Object.entries(explicitExpressions)) {
      applyExpression(name, weight);
    }

    const emotionExpressions = deriveEmotionExpressionWeights(frame);
    for (const [name, weight] of Object.entries(emotionExpressions)) {
      applyExpression(name, weight);
    }

    if (frame.blink && frame.blink > 0) {
      applyExpression('blink', frame.blink);
      const blinkSlots = manifest.morphs
        .filter((morph) => morph.role === 'blink')
        .map((morph) => morph.slot);
      applySlots(blinkSlots, frame.blink);
    }

    for (const [viseme, weight] of Object.entries(frame.visemes ?? {})) {
      applySlots(manifest.visemeAliases[normalizeName(viseme)] ?? [], weight);
    }

    return Array.from(weights.entries())
      .map(([slot, weight]) => ({ slot, weight: clampUnit(weight) }))
      .sort((left, right) => left.slot - right.slot);

  } catch (error) {
    throw new Error(`[mixAvatarMorphDeltas] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Handles the gnosis evaluate Avatar Pose workflow.
 */
export function evaluateAvatarPose(
  manifest: GnosisAvatarRuntimeManifest,
  frame: GnosisAvatarSignalFrame
): GnosisAvatarPoseResult {
  try {
    const clipName = chooseClip(frame);
    const merged = new Map<number, GnosisAvatarQuaternion>();

    for (const boneDelta of sampleProceduralAvatarClip(
      manifest,
      clipName,
      frame.timestampMs
    )) {
      addMergedBoneDelta(merged, boneDelta.slot, boneDelta.rotation);
    }

    if (frame.gazeTarget) {
      for (const boneDelta of solveLookAtBoneDeltas(
        manifest,
        frame.gazeTarget,
        frame.lookAtWeight ?? 1
      )) {
        addMergedBoneDelta(merged, boneDelta.slot, boneDelta.rotation);
      }
    }

    return {
      activeClip: clipName,
      boneDeltas: Array.from(merged.entries())
        .map(([slot, rotation]) => ({ slot, rotation }))
        .sort((left, right) => left.slot - right.slot),
      morphDeltas: mixAvatarMorphDeltas(manifest, frame),
      aura: {
        energy: clampUnit(
          (frame.emotion?.arousal ?? 0.35) * 0.6 +
            (frame.emotion?.focus ?? 0.4) * 0.4
        ),
        turbulence: clampUnit(
          (frame.emotion?.stress ?? 0) * 0.7 +
            (frame.speechState === 'speaking' ? 0.15 : 0)
        ),
        focus: clampUnit(frame.emotion?.focus ?? 0.4),
        tint: tintFromEmotion(frame.emotion),
      },
    };

  } catch (error) {
    throw new Error(`[evaluateAvatarPose] ${error instanceof Error ? error.message : String(error)}`);
  }
}
