// @a0n/.ignored_gnosis/src/ballistic-vest-policy.ts
/**
 * BallisticVest policy mirror.
 *
 * Runtime-side finite Nat mirror for
 * `open-source/gnosis-math/Gnosis/BallisticVest.lean`. The functions here
 * intentionally keep the same small arithmetic carrier: demand is
 * `mass * velocitySquared`, load is divided across `fiberDensity + 1`, reserve
 * is saturating subtraction, and ceramic shatter divides velocity-square load
 * by 10 while clearing localization.
 */

export interface KineticWave {
  readonly mass: number;
  readonly velocitySquared: number;
  readonly localization: number;
}

export interface BraidedVest {
  readonly fiberDensity: number;
  readonly bettiReserve: number;
  readonly dampingCoefficient: number;
}

export interface CeramicShatterReport {
  readonly wave: KineticWave;
  readonly distributedLoad: number;
  readonly reducedVelocity: boolean;
  readonly delocalized: boolean;
}

export interface BallisticVestPosture {
  readonly wave: KineticWave;
  readonly vest: BraidedVest;
  readonly kineticDemand: number;
  readonly erasureHeat: number;
  readonly distributedLoad: number;
  readonly residualReserve: number;
  readonly reservePositive: boolean;
  readonly denseWeave: boolean;
  readonly loadReducedByWeave: boolean;
  readonly ceramic: CeramicShatterReport;
  readonly verdict: 'admit' | 'warn' | 'collapse';
  readonly theoremLineage: readonly string[];
}

function finiteNat(value: number, name: string): number {
  if (!Number.isFinite(value) || value < 0) {
    throw new Error(`${name} must be a finite non-negative number.`);
  }
  return Math.floor(value);
}

/**
 * Handles the gnosis normalize Kinetic Wave workflow.
 */
export function normalizeKineticWave(wave: KineticWave): KineticWave {
  return {
    mass: finiteNat(wave.mass, 'wave.mass'),
    velocitySquared: finiteNat(wave.velocitySquared, 'wave.velocitySquared'),
    localization: finiteNat(wave.localization, 'wave.localization'),
  };
}

/**
 * Handles the gnosis normalize Braided Vest workflow.
 */
export function normalizeBraidedVest(vest: BraidedVest): BraidedVest {
  return {
    fiberDensity: finiteNat(vest.fiberDensity, 'vest.fiberDensity'),
    bettiReserve: finiteNat(vest.bettiReserve, 'vest.bettiReserve'),
    dampingCoefficient: finiteNat(
      vest.dampingCoefficient,
      'vest.dampingCoefficient',
    ),
  };
}

/**
 * Handles the gnosis kinetic Demand workflow.
 */
export function kineticDemand(wave: KineticWave): number {
  const normalized = normalizeKineticWave(wave);
  return normalized.mass * normalized.velocitySquared;
}

/**
 * Handles the gnosis erasure Heat workflow.
 */
export function erasureHeat(wave: KineticWave): number {
  return Math.floor(kineticDemand(wave) / 2);
}

/**
 * Handles the gnosis residual Reserve workflow.
 */
export function residualReserve(wave: KineticWave, vest: BraidedVest): number {
  const normalizedVest = normalizeBraidedVest(vest);
  return Math.max(0, normalizedVest.bettiReserve - kineticDemand(wave));
}

/**
 * Handles the gnosis distribute Force workflow.
 */
export function distributeForce(
  wave: KineticWave,
  vest: BraidedVest,
): number {
  const normalizedWave = normalizeKineticWave(wave);
  const normalizedVest = normalizeBraidedVest(vest);
  return Math.floor(
    normalizedWave.velocitySquared / (normalizedVest.fiberDensity + 1),
  );
}

/**
 * Handles the gnosis ceramic Shatter workflow.
 */
export function ceramicShatter(wave: KineticWave): KineticWave {
  const normalized = normalizeKineticWave(wave);
  return {
    ...normalized,
    localization: 0,
    velocitySquared: Math.floor(normalized.velocitySquared / 10),
  };
}

/**
 * Handles the gnosis evaluate Ballistic Vest Posture workflow.
 */
export function evaluateBallisticVestPosture(
  wave: KineticWave,
  vest: BraidedVest,
): BallisticVestPosture {
  const normalizedWave = normalizeKineticWave(wave);
  const normalizedVest = normalizeBraidedVest(vest);
  const demand = kineticDemand(normalizedWave);
  const distributedLoad = distributeForce(normalizedWave, normalizedVest);
  const reserve = residualReserve(normalizedWave, normalizedVest);
  const ceramicWave = ceramicShatter(normalizedWave);
  const ceramicDistributedLoad = distributeForce(ceramicWave, normalizedVest);
  const denseWeave = normalizedVest.fiberDensity > 0;
  const loadReducedByWeave =
    normalizedWave.velocitySquared === 0 ||
    distributedLoad < normalizedWave.velocitySquared;
  const reservePositive = reserve > 0;
  const ceramicReducedVelocity =
    normalizedWave.velocitySquared === 0 ||
    ceramicWave.velocitySquared < normalizedWave.velocitySquared;
  const ceramicDelocalized = ceramicWave.localization === 0;
  const verdict =
    !reservePositive ? 'collapse' : loadReducedByWeave ? 'admit' : 'warn';

  return {
    wave: normalizedWave,
    vest: normalizedVest,
    kineticDemand: demand,
    erasureHeat: Math.floor(demand / 2),
    distributedLoad,
    residualReserve: reserve,
    reservePositive,
    denseWeave,
    loadReducedByWeave,
    ceramic: {
      wave: ceramicWave,
      distributedLoad: ceramicDistributedLoad,
      reducedVelocity: ceramicReducedVelocity,
      delocalized: ceramicDelocalized,
    },
    verdict,
    theoremLineage: [
      'distributed_load_bounded_by_source',
      'dense_weave_models_destructive_interference',
      'betti_reserve_prevents_topological_collapse',
      'ceramic_shatter_delocalizes',
      'ceramic_shatter_enables_mesh_survival',
    ],
  };
}
