// @a0n/.ignored_gnosis/src/betti-wasm-store.ts
/**
 * Betti → WASM artifact registry store (D1 + R2 + local file registry).
 *
 * Persists `betti-wasm/1` artifacts (module bytes + manifest sidecar) into
 * the immutable registry described by the Betti→WASM artifact contract:
 *
 *   - D1 table `betti_wasm_artifacts` (sibling of `polyglot_fragments` in
 *     `polyglot-vfs.ts`), keyed by the emitter's explicit `artifactId` with a
 *     secondary `key` column = `fragmentKey(filePath, functionName)`.
 *   - R2 objects `betti/v1/<artifactId>.wasm` + `betti/v1/<artifactId>.manifest.json`.
 *   - A local file-registry mode (`storeBettiWasmArtifactToDirectory`) for
 *     dev/CLI flows where no D1/R2 bindings exist — same layout under a
 *     directory plus a deterministic `registry.json` index.
 *
 * Integrity is verified before anything is written: the manifest's `sha256`
 * must equal sha256 over `wasmBytes[..byteLength]` (the module bytes
 * EXCLUSIVE of the trailing `gnosis-betti-manifest` custom section).
 * Everything fails closed — malformed manifests, tampered bytes, and
 * truncated modules throw; they are never partially stored.
 *
 * Hashing prefers the Web Crypto API (`crypto.subtle`, edge-safe) and falls
 * back to a lazy `node:crypto` import; Node built-ins are only imported
 * lazily so the D1/R2 path stays clean for Workers.
 */
import {
  validateBettiWasmManifest,
  type BettiWasmManifest,
} from './betti-wasm-manifest';
import { fragmentKey, type D1Database } from './polyglot-vfs';

export type { D1Database } from './polyglot-vfs';

/** R2 key prefix for the immutable Betti WASM registry. */
export const BETTI_WASM_R2_PREFIX = 'betti/v1';

/** D1 table holding registered Betti WASM artifacts. */
export const BETTI_WASM_ARTIFACTS_TABLE = 'betti_wasm_artifacts';

/** R2 object key for an artifact's module bytes. */
export function bettiWasmModuleKey(artifactId: string): string {
  return `${BETTI_WASM_R2_PREFIX}/${artifactId}.wasm`;
}

/** R2 object key for an artifact's manifest sidecar. */
export function bettiWasmManifestKey(artifactId: string): string {
  return `${BETTI_WASM_R2_PREFIX}/${artifactId}.manifest.json`;
}

/**
 * Row shape of the `betti_wasm_artifacts` D1 table.
 *
 * ```sql
 * CREATE TABLE IF NOT EXISTS betti_wasm_artifacts (
 *   artifact_id TEXT PRIMARY KEY,
 *   key TEXT NOT NULL,            -- fragmentKey(file_path, function_name)
 *   file_path TEXT NOT NULL,
 *   function_name TEXT NOT NULL,
 *   language TEXT NOT NULL,
 *   tier TEXT NOT NULL,
 *   pure INTEGER NOT NULL,        -- 0 | 1
 *   manifest_json TEXT NOT NULL,
 *   r2_key TEXT,                  -- NULL when no R2 bucket was provided
 *   byte_length INTEGER NOT NULL, -- module bytes EXCLUSIVE of custom section
 *   sha256 TEXT NOT NULL,         -- 64-hex over bytes[..byte_length]
 *   created_at TEXT NOT NULL,
 *   owner_did TEXT,               -- Wave D3: artifact owner DID (earn lane)
 *   price_cents INTEGER           -- Wave D3: listing price, USD cents
 * );
 * ```
 *
 * `owner_did` / `price_cents` were added after the table shipped, so
 * {@link ensureBettiWasmArtifactsTable} both includes them in the CREATE (fresh
 * databases) and applies guarded `ALTER TABLE ... ADD COLUMN` migrations
 * (existing databases). Pre-migration rows read back as `NULL` for both.
 */
export interface BettiWasmArtifactRow {
  artifact_id: string;
  key: string;
  file_path: string;
  function_name: string;
  language: string;
  tier: string;
  pure: number;
  manifest_json: string;
  r2_key: string | null;
  byte_length: number;
  sha256: string;
  created_at: string;
  owner_did: string | null;
  price_cents: number | null;
}

/**
 * Minimal structural R2 bucket interface (matches Workers `R2Bucket` for the
 * operations used here, and is trivial to fake in tests).
 */
export interface BettiWasmR2Bucket {
  put(key: string, value: Uint8Array | string): Promise<unknown>;
  get(
    key: string
  ): Promise<{ arrayBuffer(): Promise<ArrayBuffer> } | null>;
}

function storeError(message: string): Error {
  return new Error(`betti-wasm store: ${message}`);
}

const HEX_TABLE = Array.from({ length: 256 }, (_value, index) =>
  index.toString(16).padStart(2, '0')
);

function bytesToHex(bytes: Uint8Array): string {
  let hex = '';
  for (let index = 0; index < bytes.length; index += 1) {
    hex += HEX_TABLE[bytes[index]];
  }
  return hex;
}

/**
 * sha256 as 64 lowercase hex characters. Prefers Web Crypto (edge-safe);
 * falls back to a lazy `node:crypto` import on runtimes without
 * `crypto.subtle`.
 */
export async function sha256Hex(bytes: Uint8Array): Promise<string> {
  const subtle = (
    globalThis as { crypto?: { subtle?: SubtleCrypto } }
  ).crypto?.subtle;
  if (subtle) {
    const digest = await subtle.digest(
      'SHA-256',
      // Re-slice so a view into a larger buffer hashes only its own bytes.
      bytes.slice()
    );
    return bytesToHex(new Uint8Array(digest));
  }

  const { createHash } = await import('node:crypto');
  return createHash('sha256').update(bytes).digest('hex');
}

/**
 * Validate a manifest JSON + module byte pair and verify integrity:
 * `manifest.sha256 === sha256(wasmBytes[..manifest.byteLength])`. The bytes
 * past `byteLength` are the trailing `gnosis-betti-manifest` custom section
 * and are excluded from the hash by contract. Throws on any mismatch.
 */
export async function validateBettiWasmArtifactPair(
  wasmBytes: Uint8Array,
  manifestJson: string
): Promise<BettiWasmManifest> {
  let parsed: unknown;
  try {
    parsed = JSON.parse(manifestJson);
  } catch (error) {
    throw storeError(
      `manifest is not valid JSON: ${
        error instanceof Error ? error.message : String(error)
      }`
    );
  }
  const manifest = validateBettiWasmManifest(parsed);

  if (manifest.byteLength > wasmBytes.length) {
    throw storeError(
      `manifest byteLength ${manifest.byteLength} exceeds module size ` +
        `${wasmBytes.length} — truncated or mismatched artifact pair`
    );
  }

  const digest = await sha256Hex(wasmBytes.subarray(0, manifest.byteLength));
  if (digest !== manifest.sha256) {
    throw storeError(
      `integrity failure for artifact ${manifest.artifactId}: manifest ` +
        `sha256 ${manifest.sha256} != computed ${digest} over ` +
        `bytes[..${manifest.byteLength}]`
    );
  }

  return manifest;
}

const CREATE_TABLE_SQL = `CREATE TABLE IF NOT EXISTS ${BETTI_WASM_ARTIFACTS_TABLE} (
  artifact_id TEXT PRIMARY KEY,
  key TEXT NOT NULL,
  file_path TEXT NOT NULL,
  function_name TEXT NOT NULL,
  language TEXT NOT NULL,
  tier TEXT NOT NULL,
  pure INTEGER NOT NULL,
  manifest_json TEXT NOT NULL,
  r2_key TEXT,
  byte_length INTEGER NOT NULL,
  sha256 TEXT NOT NULL,
  created_at TEXT NOT NULL,
  owner_did TEXT,
  price_cents INTEGER
)`;

const CREATE_KEY_INDEX_SQL = `CREATE INDEX IF NOT EXISTS idx_betti_wasm_artifacts_key ON ${BETTI_WASM_ARTIFACTS_TABLE}(key)`;

/**
 * Versioned column additions (Wave D3). `CREATE TABLE IF NOT EXISTS` is a
 * no-op on existing tables, so pre-existing databases need explicit
 * `ADD COLUMN` migrations. SQLite/D1 answer "duplicate column name" when the
 * column already exists — that exact failure is swallowed (idempotent
 * re-ensure); every other failure propagates.
 */
const ADD_COLUMN_MIGRATIONS_SQL = [
  `ALTER TABLE ${BETTI_WASM_ARTIFACTS_TABLE} ADD COLUMN owner_did TEXT`,
  `ALTER TABLE ${BETTI_WASM_ARTIFACTS_TABLE} ADD COLUMN price_cents INTEGER`,
] as const;

function isDuplicateColumnError(error: unknown): boolean {
  return /duplicate column/i.test(
    error instanceof Error ? error.message : String(error)
  );
}

/**
 * Create the `betti_wasm_artifacts` table (and its `key` index) if absent,
 * then apply the guarded ADD COLUMN migrations so pre-existing tables gain
 * `owner_did` / `price_cents` (existing rows keep NULL). Statements run
 * individually through `prepare().run()` so the same code works against
 * Workers D1 and structural fakes in tests.
 */
export async function ensureBettiWasmArtifactsTable(
  db: D1Database
): Promise<void> {
  await db.prepare(CREATE_TABLE_SQL).run();
  await db.prepare(CREATE_KEY_INDEX_SQL).run();
  for (const migration of ADD_COLUMN_MIGRATIONS_SQL) {
    try {
      await db.prepare(migration).run();
    } catch (error) {
      if (!isDuplicateColumnError(error)) {
        throw error;
      }
    }
  }
}

export interface StoreBettiWasmArtifactOptions {
  /** D1 database (optional — skipped with a warning when absent). */
  readonly db?: D1Database;
  /** R2 bucket (optional — skipped with a warning when absent). */
  readonly r2?: BettiWasmR2Bucket;
  /**
   * The FULL artifact bytes as written by the emitter: core module followed by
   * the trailing `gnosis-betti-manifest` custom section. Note this is NOT what
   * the integrity fields cover — `manifest.sha256`/`byteLength` hash only
   * `wasmBytes[..byteLength]`, the core module EXCLUSIVE of that trailing
   * section (a section cannot hash itself). See `verifyBettiWasmIntegrity`.
   */
  readonly wasmBytes: Uint8Array;
  /** Manifest sidecar JSON text. */
  readonly manifestJson: string;
  /** Artifact owner DID (Wave D3 earn lane; stored as `owner_did`). */
  readonly ownerDid?: string;
  /** Listing price in USD cents (Wave D3; stored as `price_cents`). */
  readonly priceCents?: number;
  /** Clock override for deterministic tests. */
  readonly now?: () => Date;
}

export interface StoreBettiWasmArtifactResult {
  readonly manifest: BettiWasmManifest;
  /** `fragmentKey(manifest.filePath, manifest.functionName)`. */
  readonly key: string;
  readonly row: BettiWasmArtifactRow;
  readonly storedToD1: boolean;
  readonly storedToR2: boolean;
  readonly r2ModuleKey: string;
  readonly r2ManifestKey: string;
  /** Non-fatal skips (e.g. no R2 bucket provided). Never hides failures. */
  readonly warnings: readonly string[];
}

/**
 * Validate + verify an artifact pair, then upsert the D1 row and put the R2
 * object pair. Validation and integrity failures throw before any write.
 * Missing `db`/`r2` handles skip the corresponding write with an explicit
 * warning in the result — never silently.
 */
export async function storeBettiWasmArtifact(
  options: StoreBettiWasmArtifactOptions
): Promise<StoreBettiWasmArtifactResult> {
  const manifest = await validateBettiWasmArtifactPair(
    options.wasmBytes,
    options.manifestJson
  );

  const key = fragmentKey(manifest.filePath, manifest.functionName);
  const r2ModuleKey = bettiWasmModuleKey(manifest.artifactId);
  const r2ManifestKey = bettiWasmManifestKey(manifest.artifactId);
  const warnings: string[] = [];

  let storedToR2 = false;
  if (options.r2) {
    await options.r2.put(r2ModuleKey, options.wasmBytes);
    await options.r2.put(r2ManifestKey, options.manifestJson);
    storedToR2 = true;
  } else {
    warnings.push(
      `no R2 bucket provided — module bytes for ${manifest.artifactId} were not uploaded (r2_key stays NULL)`
    );
  }

  const row: BettiWasmArtifactRow = {
    artifact_id: manifest.artifactId,
    key,
    file_path: manifest.filePath,
    function_name: manifest.functionName,
    language: manifest.language,
    tier: manifest.tier,
    pure: manifest.pure ? 1 : 0,
    manifest_json: options.manifestJson,
    r2_key: storedToR2 ? r2ModuleKey : null,
    byte_length: manifest.byteLength,
    sha256: manifest.sha256,
    created_at: (options.now?.() ?? new Date()).toISOString(),
    owner_did:
      options.ownerDid !== undefined && options.ownerDid.length > 0
        ? options.ownerDid
        : null,
    price_cents:
      options.priceCents !== undefined && Number.isFinite(options.priceCents)
        ? Math.max(0, Math.round(options.priceCents))
        : null,
  };

  let storedToD1 = false;
  if (options.db) {
    await ensureBettiWasmArtifactsTable(options.db);
    await options.db
      .prepare(
        `INSERT OR REPLACE INTO ${BETTI_WASM_ARTIFACTS_TABLE}
         (artifact_id, key, file_path, function_name, language, tier, pure, manifest_json, r2_key, byte_length, sha256, created_at, owner_did, price_cents)
         VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`
      )
      .bind(
        row.artifact_id,
        row.key,
        row.file_path,
        row.function_name,
        row.language,
        row.tier,
        row.pure,
        row.manifest_json,
        row.r2_key,
        row.byte_length,
        row.sha256,
        row.created_at,
        row.owner_did,
        row.price_cents
      )
      .run();
    storedToD1 = true;
  } else {
    warnings.push(
      `no D1 database provided — registry row for ${manifest.artifactId} was not persisted`
    );
  }

  return {
    manifest,
    key,
    row,
    storedToD1,
    storedToR2,
    r2ModuleKey,
    r2ManifestKey,
    warnings,
  };
}

export interface UpsertBettiWasmArtifactRowOptions {
  /** D1 database (required — this path writes the registry row only). */
  readonly db: D1Database;
  /** Manifest sidecar JSON text (validated; no module bytes are required). */
  readonly manifestJson: string;
  /** Artifact owner DID (Wave D3 earn lane; stored as `owner_did`). */
  readonly ownerDid?: string;
  /** Listing price in USD cents (Wave D3; stored as `price_cents`). */
  readonly priceCents?: number;
  /** Clock override for deterministic tests. */
  readonly now?: () => Date;
}

export interface UpsertBettiWasmArtifactRowResult {
  readonly manifest: BettiWasmManifest;
  /** `fragmentKey(manifest.filePath, manifest.functionName)`. */
  readonly key: string;
  readonly row: BettiWasmArtifactRow;
}

/**
 * Write ONLY the `betti_wasm_artifacts` D1 registry row from a manifest — no
 * R2 module bytes, no wasm-integrity check.
 *
 * This is the lighter twin of {@link storeBettiWasmArtifact} for artifacts
 * that have no emitted module to store (Tier C gg-walker envelopes). The
 * earn lane still needs a registry row carrying `owner_did` + `price_cents`
 * so the meter's owner lookup (keyed by `artifact_id`) can resolve the
 * function's owner on a cache hit. The row's `r2_key` stays NULL and its
 * `byte_length` / `sha256` are copied from the manifest for provenance;
 * because no bytes are held here, the pair is validated for shape only.
 */
export async function upsertBettiWasmArtifactRow(
  options: UpsertBettiWasmArtifactRowOptions
): Promise<UpsertBettiWasmArtifactRowResult> {
  let parsed: unknown;
  try {
    parsed = JSON.parse(options.manifestJson);
  } catch (error) {
    throw storeError(
      `manifest is not valid JSON: ${
        error instanceof Error ? error.message : String(error)
      }`
    );
  }
  const manifest = validateBettiWasmManifest(parsed);
  const key = fragmentKey(manifest.filePath, manifest.functionName);

  const row: BettiWasmArtifactRow = {
    artifact_id: manifest.artifactId,
    key,
    file_path: manifest.filePath,
    function_name: manifest.functionName,
    language: manifest.language,
    tier: manifest.tier,
    pure: manifest.pure ? 1 : 0,
    manifest_json: options.manifestJson,
    r2_key: null,
    byte_length: manifest.byteLength,
    sha256: manifest.sha256,
    created_at: (options.now?.() ?? new Date()).toISOString(),
    owner_did:
      options.ownerDid !== undefined && options.ownerDid.length > 0
        ? options.ownerDid
        : null,
    price_cents:
      options.priceCents !== undefined && Number.isFinite(options.priceCents)
        ? Math.max(0, Math.round(options.priceCents))
        : null,
  };

  await ensureBettiWasmArtifactsTable(options.db);
  await options.db
    .prepare(
      `INSERT OR REPLACE INTO ${BETTI_WASM_ARTIFACTS_TABLE}
       (artifact_id, key, file_path, function_name, language, tier, pure, manifest_json, r2_key, byte_length, sha256, created_at, owner_did, price_cents)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`
    )
    .bind(
      row.artifact_id,
      row.key,
      row.file_path,
      row.function_name,
      row.language,
      row.tier,
      row.pure,
      row.manifest_json,
      row.r2_key,
      row.byte_length,
      row.sha256,
      row.created_at,
      row.owner_did,
      row.price_cents
    )
    .run();

  return { manifest, key, row };
}

export interface LoadBettiWasmArtifactOptions {
  readonly db: D1Database;
  /** R2 bucket; when present (and the row has an `r2_key`) bytes are fetched + re-verified. */
  readonly r2?: BettiWasmR2Bucket;
  /** Exactly one of `artifactId` or `key` must be provided. */
  readonly artifactId?: string;
  readonly key?: string;
}

export interface LoadedBettiWasmArtifact {
  readonly row: BettiWasmArtifactRow;
  readonly manifest: BettiWasmManifest;
  /** Present only when an R2 bucket was provided and the row has an `r2_key`. */
  readonly wasmBytes?: Uint8Array;
}

/**
 * Load a registered artifact by `artifactId` or fragment `key` (newest row
 * wins for a key). When an R2 bucket is provided the module bytes are
 * fetched and re-verified against the row's sha256 — a corrupt or missing
 * object throws rather than returning bad bytes. Returns `null` when no row
 * matches.
 */
export async function loadBettiWasmArtifact(
  options: LoadBettiWasmArtifactOptions
): Promise<LoadedBettiWasmArtifact | null> {
  const hasArtifactId =
    typeof options.artifactId === 'string' && options.artifactId.length > 0;
  const hasKey = typeof options.key === 'string' && options.key.length > 0;
  if (hasArtifactId === hasKey) {
    throw storeError(
      'load requires exactly one selector: artifactId or key'
    );
  }

  const rawRow = hasArtifactId
    ? await options.db
        .prepare(
          `SELECT * FROM ${BETTI_WASM_ARTIFACTS_TABLE} WHERE artifact_id = ?`
        )
        .bind(options.artifactId)
        .first<BettiWasmArtifactRow>()
    : await options.db
        .prepare(
          `SELECT * FROM ${BETTI_WASM_ARTIFACTS_TABLE} WHERE key = ? ORDER BY created_at DESC LIMIT 1`
        )
        .bind(options.key)
        .first<BettiWasmArtifactRow>();

  if (!rawRow) {
    return null;
  }

  // Backward compat: rows read from a pre-D3 table (never re-ensured) have
  // no owner/price columns at all — normalize to explicit NULLs.
  const row: BettiWasmArtifactRow = {
    ...rawRow,
    owner_did: rawRow.owner_did ?? null,
    price_cents: rawRow.price_cents ?? null,
  };

  const manifest = validateBettiWasmManifest(JSON.parse(row.manifest_json));

  if (!options.r2 || row.r2_key === null) {
    return { row, manifest };
  }

  const object = await options.r2.get(row.r2_key);
  if (!object) {
    throw storeError(
      `registry row ${row.artifact_id} points at missing R2 object '${row.r2_key}'`
    );
  }
  const wasmBytes = new Uint8Array(await object.arrayBuffer());
  if (wasmBytes.length < row.byte_length) {
    throw storeError(
      `R2 object '${row.r2_key}' is ${wasmBytes.length} byte(s), shorter than the recorded byte_length ${row.byte_length}`
    );
  }
  const digest = await sha256Hex(wasmBytes.subarray(0, row.byte_length));
  if (digest !== row.sha256) {
    throw storeError(
      `integrity failure loading ${row.artifact_id} from '${row.r2_key}': ` +
        `stored sha256 ${row.sha256} != computed ${digest}`
    );
  }

  return { row, manifest, wasmBytes };
}

// ── Local file registry (dev/CLI mode — no D1/R2 bindings) ──────────────────

/** Index entry inside a file registry's `registry.json`. */
export interface BettiWasmFileRegistryEntry {
  readonly artifactId: string;
  readonly key: string;
  readonly filePath: string;
  readonly functionName: string;
  readonly language: string;
  readonly tier: string;
  readonly pure: boolean;
  readonly byteLength: number;
  readonly sha256: string;
  readonly createdAt: string;
  /** Paths relative to the registry root (mirror the R2 key layout). */
  readonly wasmPath: string;
  readonly manifestPath: string;
  /** Wave D3 (optional — only present when supplied at store time). */
  readonly ownerDid?: string;
  readonly priceCents?: number;
}

export interface BettiWasmFileRegistryIndex {
  readonly version: 1;
  readonly artifacts: readonly BettiWasmFileRegistryEntry[];
}

export interface StoreBettiWasmArtifactToDirectoryOptions {
  /** Registry root directory (created if absent). */
  readonly dir: string;
  readonly wasmBytes: Uint8Array;
  readonly manifestJson: string;
  /** Artifact owner DID (Wave D3; recorded on the index entry). */
  readonly ownerDid?: string;
  /** Listing price in USD cents (Wave D3; recorded on the index entry). */
  readonly priceCents?: number;
  /** Clock override for deterministic tests. */
  readonly now?: () => Date;
}

export interface StoreBettiWasmArtifactToDirectoryResult {
  readonly manifest: BettiWasmManifest;
  readonly entry: BettiWasmFileRegistryEntry;
  readonly wasmPath: string;
  readonly manifestPath: string;
  readonly registryPath: string;
}

/**
 * File-registry twin of {@link storeBettiWasmArtifact} for local/dev flows:
 * writes `<dir>/betti/v1/<artifactId>.wasm` + `.manifest.json` (the exact R2
 * key layout) and maintains a deterministic `<dir>/registry.json` index
 * sorted by artifactId. Same fail-closed validation core; Node built-ins are
 * imported lazily so this module stays edge-loadable.
 */
export async function storeBettiWasmArtifactToDirectory(
  options: StoreBettiWasmArtifactToDirectoryOptions
): Promise<StoreBettiWasmArtifactToDirectoryResult> {
  const manifest = await validateBettiWasmArtifactPair(
    options.wasmBytes,
    options.manifestJson
  );

  const [{ mkdir, readFile, writeFile }, path] = await Promise.all([
    import('node:fs/promises'),
    import('node:path'),
  ]);

  const relativeWasmPath = bettiWasmModuleKey(manifest.artifactId);
  const relativeManifestPath = bettiWasmManifestKey(manifest.artifactId);
  const wasmPath = path.join(options.dir, relativeWasmPath);
  const manifestPath = path.join(options.dir, relativeManifestPath);
  const registryPath = path.join(options.dir, 'registry.json');

  await mkdir(path.dirname(wasmPath), { recursive: true });
  await writeFile(wasmPath, options.wasmBytes);
  await writeFile(manifestPath, options.manifestJson, 'utf8');

  const entry: BettiWasmFileRegistryEntry = {
    artifactId: manifest.artifactId,
    key: fragmentKey(manifest.filePath, manifest.functionName),
    filePath: manifest.filePath,
    functionName: manifest.functionName,
    language: manifest.language,
    tier: manifest.tier,
    pure: manifest.pure,
    byteLength: manifest.byteLength,
    sha256: manifest.sha256,
    createdAt: (options.now?.() ?? new Date()).toISOString(),
    wasmPath: relativeWasmPath,
    manifestPath: relativeManifestPath,
    // Wave D3 fields appear only when supplied — absent keys keep older
    // registries byte-identical on idempotent re-stores.
    ...(options.ownerDid !== undefined && options.ownerDid.length > 0
      ? { ownerDid: options.ownerDid }
      : {}),
    ...(options.priceCents !== undefined && Number.isFinite(options.priceCents)
      ? { priceCents: Math.max(0, Math.round(options.priceCents)) }
      : {}),
  };

  let existing: BettiWasmFileRegistryEntry[] = [];
  try {
    const raw = JSON.parse(await readFile(registryPath, 'utf8')) as {
      version?: unknown;
      artifacts?: unknown;
    };
    if (raw.version !== 1 || !Array.isArray(raw.artifacts)) {
      throw storeError(
        `existing registry index at ${registryPath} has an unknown shape — refusing to overwrite`
      );
    }
    existing = raw.artifacts as BettiWasmFileRegistryEntry[];
  } catch (error) {
    if (
      !(
        error &&
        typeof error === 'object' &&
        (error as NodeJS.ErrnoException).code === 'ENOENT'
      )
    ) {
      throw error;
    }
  }

  const merged = new Map<string, BettiWasmFileRegistryEntry>(
    existing.map((item) => [item.artifactId, item])
  );
  merged.set(entry.artifactId, entry);
  const index: BettiWasmFileRegistryIndex = {
    version: 1,
    artifacts: [...merged.values()].sort((left, right) =>
      left.artifactId.localeCompare(right.artifactId)
    ),
  };
  await writeFile(registryPath, `${JSON.stringify(index, null, 2)}\n`, 'utf8');

  return { manifest, entry, wasmPath, manifestPath, registryPath };
}
