// @a0n/.ignored_gnosis/src/betti-wasm-cli.ts
/**
 * `gnode wasm-store` / `gnode wasm-codegen` — Betti → WASM registry CLI.
 *
 * wasm-store shells the Betti/polyglot emitter (`monster`/`gnexec`,
 * `--emit-wasm-out` writes the module + `.manifest.json` sidecar pair),
 * validates + integrity-checks the pair, and stores it into a LOCAL FILE
 * REGISTRY (`--out-dir`): `<dir>/betti/v1/<artifactId>.{wasm,manifest.json}`
 * plus a deterministic `<dir>/registry.json` index.
 *
 * HONESTY NOTE — no local D1/R2 here: this package has no local D1 shim
 * (no better-sqlite3/miniflare dependency), so the CLI does not pretend to
 * write D1 or R2. Real D1/R2 persistence is the binding-injected path:
 * Workers-side callers pass their `env.DB` / `env.BUCKET` bindings to
 * `storeBettiWasmArtifact` in `betti-wasm-store.ts`. The file registry's
 * `betti/v1/` layout mirrors the R2 key layout exactly, so promoting a local
 * registry to R2 is a plain object copy.
 *
 * wasm-codegen turns an artifacts directory into a Worker-side
 * `betti-artifacts.ts` CompiledWasm module map (see `betti-wasm-codegen.ts`).
 */
import { execFile } from 'node:child_process';
import { existsSync } from 'node:fs';
import { mkdtemp, readFile, rm } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { delimiter, dirname, join, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import { promisify } from 'node:util';
import { generateBettiArtifactsModule } from './betti-wasm-codegen';
import { storeBettiWasmArtifactToDirectory } from './betti-wasm-store';

const execFileAsync = promisify(execFile);

const HERE = dirname(fileURLToPath(import.meta.url));

/**
 * The gnosis package root, resolved by walking up from this module until a
 * sibling `polyglot/` directory appears. This module runs from `src/` (tsx
 * path) AND from `dist/gnode/` (the bundled gnode driver), so a fixed
 * `../polyglot` offset would break one of the two.
 */
function resolvePolyglotTargetDir(): string | null {
  let current = HERE;
  for (let depth = 0; depth < 5; depth += 1) {
    const candidate = join(current, 'polyglot', 'target');
    if (existsSync(candidate)) {
      return candidate;
    }
    const parent = dirname(current);
    if (parent === current) {
      break;
    }
    current = parent;
  }
  return null;
}

/**
 * Resolve the Betti emitter binary. Mirrors x-gnosis
 * `resolveMonsterEmitterBinary`: `$GNOSIS_MONSTER_BIN` → polyglot
 * `target/release/monster` → `target/release/gnexec` → `target/debug/gnexec`
 * → `monster` on PATH. Returns `null` when nothing resolves (callers fail
 * closed).
 */
export function resolveBettiEmitterBinary(): string | null {
  const fromEnv = process.env.GNOSIS_MONSTER_BIN;
  if (fromEnv && fromEnv.length > 0 && existsSync(fromEnv)) {
    return fromEnv;
  }

  const targetDir = resolvePolyglotTargetDir();
  if (targetDir !== null) {
    for (const candidate of [
      join(targetDir, 'release', 'monster'),
      join(targetDir, 'release', 'gnexec'),
      join(targetDir, 'debug', 'gnexec'),
    ]) {
      if (existsSync(candidate)) {
        return candidate;
      }
    }
  }

  const pathEnv = process.env.PATH ?? '';
  for (const dir of pathEnv.split(delimiter)) {
    if (dir.length > 0 && existsSync(join(dir, 'monster'))) {
      return join(dir, 'monster');
    }
  }

  return null;
}

export interface EmittedBettiWasmPair {
  readonly wasmBytes: Uint8Array;
  readonly manifestJson: string;
}

/**
 * Shell the emitter for one artifact pair:
 * `<binary> <sourceFile> --export <fn> --emit-wasm-out <tmp>/module.wasm`,
 * then read module + sidecar. Emitter declines throw with the emitter's
 * stderr — no fallback.
 */
export async function emitBettiWasmPair(options: {
  readonly sourceFile: string;
  readonly exportName: string;
  readonly emitterBinary?: string;
}): Promise<EmittedBettiWasmPair> {
  const binary = options.emitterBinary ?? resolveBettiEmitterBinary();
  if (!binary) {
    throw new Error(
      'wasm-store: no Betti emitter binary resolved. Tried $GNOSIS_MONSTER_BIN, ' +
        'polyglot/target/{release/monster, release/gnexec, debug/gnexec}, and ' +
        '`monster` on PATH. Build one with `cargo build --bin gnexec --release` ' +
        'in open-source/gnosis/polyglot.'
    );
  }

  const workDir = await mkdtemp(join(tmpdir(), 'betti-wasm-store-'));
  try {
    const modulePath = join(workDir, 'module.wasm');
    try {
      await execFileAsync(binary, [
        options.sourceFile,
        '--export',
        options.exportName,
        '--emit-wasm-out',
        modulePath,
      ]);
    } catch (error) {
      const stderr =
        error && typeof error === 'object' && 'stderr' in error
          ? String((error as { stderr: unknown }).stderr).trim()
          : '';
      throw new Error(
        `wasm-store: emitter declined '${options.exportName}' in ` +
          `'${options.sourceFile}': ${
            stderr || (error instanceof Error ? error.message : String(error))
          }`
      );
    }

    const [wasmBytes, manifestJson] = await Promise.all([
      readFile(modulePath),
      readFile(join(workDir, 'module.manifest.json'), 'utf8'),
    ]);
    return { wasmBytes: new Uint8Array(wasmBytes), manifestJson };
  } finally {
    await rm(workDir, { recursive: true, force: true });
  }
}

const WASM_STORE_USAGE = [
  'Usage: gnode wasm-store <sourceFile> --export <fn> --out-dir <dir>',
  '       [--emitter <binary>] [--owner-did <did>] [--price-cents <n>]',
  '',
  'Emits a betti-wasm/1 artifact pair via the Betti emitter and stores it in',
  'a local file registry: <dir>/betti/v1/<artifactId>.{wasm,manifest.json} +',
  '<dir>/registry.json (deterministic index, sorted by artifactId).',
  '',
  '--owner-did / --price-cents (Wave D3) record the artifact owner and its',
  'listing price on the registry entry (D1 columns owner_did / price_cents',
  'on the binding-injected path) — cache-served replays credit the owner.',
  '',
  'The file layout mirrors the R2 key layout (betti/v1/...). Real D1/R2',
  'persistence is binding-injected: call storeBettiWasmArtifact from',
  '@a0n/gnosis/betti-wasm-store with Workers env.DB / bucket bindings.',
].join('\n');

const WASM_CODEGEN_USAGE = [
  'Usage: gnode wasm-codegen <artifacts-dir> --out <file>',
  '',
  'Scans <artifacts-dir> for .wasm + .manifest.json pairs (flat emitter',
  'output or a wasm-store file registry) and generates a deterministic',
  'betti-artifacts.ts CompiledWasm module map (plus wasm/ copies and a',
  'wasm-modules.d.ts ambient declaration) at <file>.',
].join('\n');

interface WasmStoreArgs {
  readonly sourceFile: string;
  readonly exportName: string;
  readonly outDir: string;
  readonly emitterBinary?: string;
  readonly ownerDid?: string;
  readonly priceCents?: number;
}

function parseWasmStoreArgs(argv: readonly string[]): WasmStoreArgs {
  const sourceFile = argv[0];
  if (!sourceFile || sourceFile.startsWith('--')) {
    throw new Error(WASM_STORE_USAGE);
  }

  let exportName: string | undefined;
  let outDir: string | undefined;
  let emitterBinary: string | undefined;
  let ownerDid: string | undefined;
  let priceCents: number | undefined;

  for (let index = 1; index < argv.length; index += 1) {
    const flag = argv[index];
    if (flag === '--export') {
      exportName = argv[index + 1];
      index += 1;
      continue;
    }
    if (flag === '--out-dir') {
      outDir = argv[index + 1];
      index += 1;
      continue;
    }
    if (flag === '--emitter') {
      emitterBinary = argv[index + 1];
      index += 1;
      continue;
    }
    if (flag === '--owner-did') {
      ownerDid = argv[index + 1];
      index += 1;
      if (!ownerDid || ownerDid.startsWith('--')) {
        throw new Error(`--owner-did requires a value.\n\n${WASM_STORE_USAGE}`);
      }
      continue;
    }
    if (flag === '--price-cents') {
      const raw = argv[index + 1];
      index += 1;
      const parsedCents = Number(raw);
      if (
        raw === undefined ||
        !Number.isFinite(parsedCents) ||
        parsedCents < 0
      ) {
        throw new Error(
          `--price-cents requires a non-negative number, got '${raw}'.\n\n${WASM_STORE_USAGE}`
        );
      }
      priceCents = Math.round(parsedCents);
      continue;
    }
    throw new Error(`Unknown flag '${flag}'.\n\n${WASM_STORE_USAGE}`);
  }

  if (!exportName || !outDir) {
    throw new Error(WASM_STORE_USAGE);
  }

  return {
    sourceFile: resolve(process.cwd(), sourceFile),
    exportName,
    outDir: resolve(process.cwd(), outDir),
    ...(emitterBinary
      ? { emitterBinary: resolve(process.cwd(), emitterBinary) }
      : {}),
    ...(ownerDid !== undefined ? { ownerDid } : {}),
    ...(priceCents !== undefined ? { priceCents } : {}),
  };
}

/**
 * Handles the gnosis run Wasm Store Command workflow.
 */
export async function runWasmStoreCommand(
  argv: readonly string[]
): Promise<number> {
  let parsed: WasmStoreArgs;
  try {
    parsed = parseWasmStoreArgs(argv);
  } catch (error) {
    process.stderr.write(
      `${error instanceof Error ? error.message : String(error)}\n`
    );
    return 1;
  }

  try {
    const pair = await emitBettiWasmPair({
      sourceFile: parsed.sourceFile,
      exportName: parsed.exportName,
      ...(parsed.emitterBinary ? { emitterBinary: parsed.emitterBinary } : {}),
    });
    const stored = await storeBettiWasmArtifactToDirectory({
      dir: parsed.outDir,
      wasmBytes: pair.wasmBytes,
      manifestJson: pair.manifestJson,
      ...(parsed.ownerDid !== undefined ? { ownerDid: parsed.ownerDid } : {}),
      ...(parsed.priceCents !== undefined
        ? { priceCents: parsed.priceCents }
        : {}),
    });
    process.stdout.write(
      `${JSON.stringify(
        {
          artifactId: stored.manifest.artifactId,
          key: stored.entry.key,
          abi: stored.manifest.abi,
          tier: stored.manifest.tier,
          byteLength: stored.manifest.byteLength,
          sha256: stored.manifest.sha256,
          wasmPath: stored.wasmPath,
          manifestPath: stored.manifestPath,
          registryPath: stored.registryPath,
          ...(stored.entry.ownerDid !== undefined
            ? { ownerDid: stored.entry.ownerDid }
            : {}),
          ...(stored.entry.priceCents !== undefined
            ? { priceCents: stored.entry.priceCents }
            : {}),
        },
        null,
        2
      )}\n`
    );
    return 0;
  } catch (error) {
    process.stderr.write(
      `${error instanceof Error ? error.message : String(error)}\n`
    );
    return 1;
  }
}

interface WasmCodegenArgs {
  readonly artifactsDir: string;
  readonly outFile: string;
}

function parseWasmCodegenArgs(argv: readonly string[]): WasmCodegenArgs {
  const artifactsDir = argv[0];
  if (!artifactsDir || artifactsDir.startsWith('--')) {
    throw new Error(WASM_CODEGEN_USAGE);
  }

  let outFile: string | undefined;
  for (let index = 1; index < argv.length; index += 1) {
    const flag = argv[index];
    if (flag === '--out') {
      outFile = argv[index + 1];
      index += 1;
      continue;
    }
    throw new Error(`Unknown flag '${flag}'.\n\n${WASM_CODEGEN_USAGE}`);
  }

  if (!outFile) {
    throw new Error(WASM_CODEGEN_USAGE);
  }

  return {
    artifactsDir: resolve(process.cwd(), artifactsDir),
    outFile: resolve(process.cwd(), outFile),
  };
}

/**
 * Handles the gnosis run Wasm Codegen Command workflow.
 */
export async function runWasmCodegenCommand(
  argv: readonly string[]
): Promise<number> {
  let parsed: WasmCodegenArgs;
  try {
    parsed = parseWasmCodegenArgs(argv);
  } catch (error) {
    process.stderr.write(
      `${error instanceof Error ? error.message : String(error)}\n`
    );
    return 1;
  }

  try {
    const result = await generateBettiArtifactsModule(parsed);
    process.stdout.write(
      `${JSON.stringify(
        {
          outFile: result.outFile,
          wasmDir: result.wasmDir,
          declarationFile: result.declarationFile,
          artifacts: result.artifacts.map((artifact) => ({
            artifactId: artifact.artifactId,
            abi: artifact.manifest.abi,
            wasm: artifact.wasmRelativePath,
          })),
        },
        null,
        2
      )}\n`
    );
    return 0;
  } catch (error) {
    process.stderr.write(
      `${error instanceof Error ? error.message : String(error)}\n`
    );
    return 1;
  }
}
