// @a0n/.ignored_gnosis/src/betti-wasm-codegen.ts
/**
 * Betti → WASM worker-side artifact map codegen.
 *
 * Scans a directory of emitted artifact pairs (`<name>.wasm` +
 * `<name>.manifest.json`, in either the emitter's flat layout or the file
 * registry's `betti/v1/` layout), validates + integrity-checks every pair,
 * and generates a `betti-artifacts.ts` module for a consuming Worker:
 *
 * ```ts
 * import m_<artifactId> from './wasm/<artifactId>.wasm';
 * export const bettiModules = new Map<string, WebAssembly.Module>([...]);
 * export const bettiManifests = [...] as const;
 * ```
 *
 * The `.wasm` files are copied into a `wasm/` subdirectory next to the
 * generated module so the Workers `[[rules]] type = "CompiledWasm"` rule
 * turns each import into a deploy-time pre-compiled `WebAssembly.Module`
 * (the vcpu-tick pattern) — no runtime compile at the edge. A
 * `wasm-modules.d.ts` ambient declaration is emitted alongside so the
 * generated imports typecheck.
 *
 * Output is deterministic: artifacts are sorted by artifactId and nothing
 * time- or path-order-dependent is written. Duplicate artifactIds with
 * identical sha256 collapse to one entry; conflicting duplicates throw.
 */
import {
  validateBettiWasmManifest,
  type BettiWasmManifest,
} from './betti-wasm-manifest';
import { validateBettiWasmArtifactPair } from './betti-wasm-store';

export const BETTI_ARTIFACTS_MODULE_HEADER =
  '// GENERATED by `gnode wasm-codegen` (open-source/gnosis/src/betti-wasm-codegen.ts). Do not edit.';

export interface GenerateBettiArtifactsModuleOptions {
  /** Directory containing emitted `.wasm` + `.manifest.json` pairs. */
  readonly artifactsDir: string;
  /** Output path of the generated `betti-artifacts.ts` module. */
  readonly outFile: string;
}

export interface GeneratedBettiArtifact {
  readonly artifactId: string;
  readonly manifest: BettiWasmManifest;
  /** Absolute source `.wasm` path inside `artifactsDir`. */
  readonly sourceWasmPath: string;
  /** Copied module path relative to the generated module (`wasm/<id>.wasm`). */
  readonly wasmRelativePath: string;
}

export interface GenerateBettiArtifactsModuleResult {
  readonly outFile: string;
  readonly wasmDir: string;
  readonly declarationFile: string;
  readonly artifacts: readonly GeneratedBettiArtifact[];
}

function codegenError(message: string): Error {
  return new Error(`betti-wasm codegen: ${message}`);
}

/**
 * Render the generated module text from validated manifests (sorted by
 * artifactId before rendering — callers may pass any order). Pure string
 * assembly; exported for direct determinism tests.
 */
export function renderBettiArtifactsModule(
  manifests: readonly BettiWasmManifest[]
): string {
  const sorted = [...manifests].sort((left, right) =>
    left.artifactId.localeCompare(right.artifactId)
  );

  const importLines = sorted.map(
    (manifest) =>
      `import m_${manifest.artifactId} from './wasm/${manifest.artifactId}.wasm';`
  );
  const mapEntries = sorted.map(
    (manifest) =>
      `  ['${manifest.artifactId}', m_${manifest.artifactId} as unknown as WebAssembly.Module],`
  );
  const manifestLiterals = sorted.map((manifest) => {
    const json = JSON.stringify(manifest, null, 2)
      .split('\n')
      .map((line) => `  ${line}`)
      .join('\n');
    return `${json},`;
  });

  return [
    BETTI_ARTIFACTS_MODULE_HEADER,
    '//',
    '// CompiledWasm artifact map for Betti-emitted `betti-wasm/1` modules. Each',
    '// import resolves to a deploy-time pre-compiled WebAssembly.Module under the',
    '// Workers `[[rules]] type = "CompiledWasm"` rule — never a runtime compile.',
    '',
    ...importLines,
    '',
    'export const bettiModules = new Map<string, WebAssembly.Module>([',
    ...mapEntries,
    ']);',
    '',
    'export const bettiManifests = [',
    ...manifestLiterals,
    '] as const;',
    '',
  ].join('\n');
}

const WASM_MODULE_DECLARATION = [
  BETTI_ARTIFACTS_MODULE_HEADER,
  '//',
  '// Ambient declaration for CompiledWasm imports: on Cloudflare Workers a',
  '// `.wasm` import is a deploy-time pre-compiled WebAssembly.Module.',
  "declare module '*.wasm' {",
  '  const wasmModule: WebAssembly.Module;',
  '  export default wasmModule;',
  '}',
  '',
].join('\n');

/**
 * Scan `artifactsDir` for artifact pairs, validate each (manifest schema +
 * sha256 integrity over `bytes[..byteLength]`), copy modules into
 * `<outDir>/wasm/<artifactId>.wasm`, and write the generated
 * `betti-artifacts.ts` + `wasm-modules.d.ts`. Fails closed on missing
 * sidecars, malformed manifests, tampered bytes, and conflicting duplicate
 * artifactIds.
 */
export async function generateBettiArtifactsModule(
  options: GenerateBettiArtifactsModuleOptions
): Promise<GenerateBettiArtifactsModuleResult> {
  const [{ mkdir, readdir, readFile, writeFile }, path] = await Promise.all([
    import('node:fs/promises'),
    import('node:path'),
  ]);

  async function collectWasmFiles(dir: string): Promise<string[]> {
    const entries = await readdir(dir, { withFileTypes: true });
    const files: string[] = [];
    for (const entry of entries) {
      const fullPath = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        files.push(...(await collectWasmFiles(fullPath)));
      } else if (entry.isFile() && entry.name.endsWith('.wasm')) {
        files.push(fullPath);
      }
    }
    return files.sort();
  }

  const wasmFiles = await collectWasmFiles(options.artifactsDir);
  if (wasmFiles.length === 0) {
    throw codegenError(
      `no .wasm artifacts found under '${options.artifactsDir}'`
    );
  }

  const byArtifactId = new Map<
    string,
    { manifest: BettiWasmManifest; wasmBytes: Uint8Array; sourcePath: string }
  >();

  for (const wasmFile of wasmFiles) {
    const sidecarPath = `${wasmFile.slice(0, -'.wasm'.length)}.manifest.json`;
    let manifestJson: string;
    try {
      manifestJson = await readFile(sidecarPath, 'utf8');
    } catch {
      throw codegenError(
        `module '${wasmFile}' has no manifest sidecar '${sidecarPath}'`
      );
    }
    const wasmBytes = new Uint8Array(await readFile(wasmFile));
    const manifest = await validateBettiWasmArtifactPair(
      wasmBytes,
      manifestJson
    );

    const existing = byArtifactId.get(manifest.artifactId);
    if (existing) {
      if (existing.manifest.sha256 !== manifest.sha256) {
        throw codegenError(
          `conflicting duplicates for artifactId ${manifest.artifactId}: ` +
            `'${existing.sourcePath}' (sha256 ${existing.manifest.sha256}) vs ` +
            `'${wasmFile}' (sha256 ${manifest.sha256})`
        );
      }
      continue;
    }
    byArtifactId.set(manifest.artifactId, {
      manifest,
      wasmBytes,
      sourcePath: wasmFile,
    });
  }

  const sorted = [...byArtifactId.values()].sort((left, right) =>
    left.manifest.artifactId.localeCompare(right.manifest.artifactId)
  );

  const outDir = path.dirname(options.outFile);
  const wasmDir = path.join(outDir, 'wasm');
  await mkdir(wasmDir, { recursive: true });

  const artifacts: GeneratedBettiArtifact[] = [];
  for (const item of sorted) {
    const wasmRelativePath = `wasm/${item.manifest.artifactId}.wasm`;
    await writeFile(path.join(outDir, wasmRelativePath), item.wasmBytes);
    artifacts.push({
      artifactId: item.manifest.artifactId,
      manifest: item.manifest,
      sourceWasmPath: item.sourcePath,
      wasmRelativePath,
    });
  }

  const declarationFile = path.join(outDir, 'wasm-modules.d.ts');
  await writeFile(
    options.outFile,
    renderBettiArtifactsModule(sorted.map((item) => item.manifest)),
    'utf8'
  );
  await writeFile(declarationFile, WASM_MODULE_DECLARATION, 'utf8');

  return { outFile: options.outFile, wasmDir, declarationFile, artifacts };
}
