// @a0n/.ignored_gnosis/src/bracketed-space.ts
import { GNOSTIC, fib } from './math';

/**
 * A rational number represented as a pair of BigInts to maintain exact precision.
 */
export interface Q {
  num: bigint;
  den: bigint;
}

/**
 * A Bracketed Real [lower, upper].
 * Instead of a single 'number', every value is a rational interval.
 */
export class QBracket {
  constructor(public readonly lower: Q, public readonly upper: Q) {
    if (lower.num * upper.den > upper.num * lower.den) {
      throw new Error('Invalid bracket: lower > upper');
    }
  }

  static fromBigInt(n: bigint): QBracket {
    const q = { num: n, den: 1n };
    return new QBracket(q, q);
  }

  get width(): Q {
    return {
      num: this.upper.num * this.lower.den - this.lower.num * this.upper.den,
      den: this.lower.den * this.upper.den
    };
  }

  /**
   * The "Sliver" (+1) in this bracketed space is the width of the gap.
   */
  get sliver(): number {
    const w = this.width;
    return Number(w.num) / Number(w.den);
  }

  toNumber(): number {
    const midpointNum =
      this.lower.num * this.upper.den + this.upper.num * this.lower.den;
    const midpointDen = 2n * this.lower.den * this.upper.den;
    return Number(midpointNum) / Number(midpointDen);
  }

  add(other: QBracket): QBracket {
    const l = {
      num: this.lower.num * other.lower.den + other.lower.num * this.lower.den,
      den: this.lower.den * other.lower.den
    };
    const u = {
      num: this.upper.num * other.upper.den + other.upper.num * this.upper.den,
      den: this.upper.den * other.upper.den
    };
    return new QBracket(l, u);
  }

  /**
   * Narrow the local bracket (consuming computational energy).
   * Each step n costs exactly n 'bule' to compute from step 0.
   */
  refine(n: number, budget: number): { bracket: QBracket; cost: number } {
    const cost = n; // Linear cost model: depth n costs n bule.
    if (cost > budget) {
      console.warn(`[Gnosis] Refinement depth ${n} exceeds budget ${budget}. Clamping.`);
      const maxN = Math.floor(budget);
      return { bracket: getPhiTowerStep(maxN), cost: maxN };
    }
    return { bracket: getPhiTowerStep(n), cost };
  }
}

/**
 * The Phi Tower: Approximating PHI using Fibonacci ratios.
 * Every step n provides a tighter bracket.
 */
export function getPhiTowerStep(n: number): QBracket {
  const n0 = n + 2;
  const f1 = BigInt(fib(2 * n0));
  const f0 = BigInt(fib(2 * n0 - 1));
  const f2 = BigInt(fib(2 * n0 + 1));

  // phi is between F(2n)/F(2n-1) and F(2n+1)/F(2n)
  return new QBracket(
    { num: f1, den: f0 },
    { num: f2, den: f1 }
  );
}

/**
 * Comparison tool.
 */
export function comparePhi(n: number) {
  const bracket = getPhiTowerStep(n);
  const mathPhi = (1 + Math.sqrt(5)) / 2;
  
  return {
    n,
    bracket: `[${bracket.lower.num}/${bracket.lower.den}, ${bracket.upper.num}/${bracket.upper.den}]`,
    width: bracket.sliver,
    mathPhi,
    diff: Math.abs(bracket.toNumber() - mathPhi),
    isInside: mathPhi >= Number(bracket.lower.num)/Number(bracket.lower.den) && 
              mathPhi <= Number(bracket.upper.num)/Number(bracket.upper.den)
  };
}
