// @a0n/.ignored_gnosis/src/brunnian.ts
/**
 * Brunnian Coupling Detector
 *
 * Detects emergent n-wise coupling that pairwise tests miss.
 * Based on brunnian_needs_three theorem (KnottyProblems.lean):
 *   pairwise analysis is blind to higher-order failures.
 *
 * If beta1(A+B+C) > beta1(A+B) + beta1(B+C) + beta1(A+C),
 * there exists an emergent knot -- a failure that only manifests
 * when all components are composed.
 *
 * Every detected Brunnian coupling becomes:
 *   1. A test case (n-wise composition test)
 *   2. A rejection record (void boundary for buleyean-rl)
 *   3. A scanner finding (BRUNNIAN_COUPLING rule)
 */

import { BettyCompiler } from './betty/compiler';

// ── Types ────────────────────────────────────────────────────

export interface TopologyModule {
  id: string;
  source: string;
  beta1?: number;
}

export interface BrunnianResult {
  /** Which modules are involved in the emergent coupling */
  modules: string[];
  /** Beta1 of the full n-wise composition */
  composedBeta1: number;
  /** Sum of all pairwise beta1 values */
  pairwiseBeta1Sum: number;
  /** The emergent coupling strength (composed - pairwise sum) */
  emergentStrength: number;
  /** Whether this is a genuine Brunnian coupling */
  isBrunnian: boolean;
}

export interface BrunnianTestCase {
  /** Human-readable description */
  description: string;
  /** Module IDs to compose */
  composition: string[];
  /** Expected failure mode */
  expectedFailure: 'deadlock' | 'race' | 'starvation' | 'unknown';
  /** The emergent beta1 */
  emergentBeta1: number;
}

export interface VoidRejection {
  /** Rejection source */
  source: 'brunnian-detector';
  /** What was rejected */
  composition: string[];
  /** Why it was rejected */
  reason: string;
  /** Numerical strength of the rejection signal */
  weight: number;
  /** Timestamp */
  timestamp: string;
}

// ── Beta1 Computation ────────────────────────────────────────

/**
 * Compute beta1 for a single topology source.
 */
function computeBeta1(source: string): number {
  const compiler = new BettyCompiler();
  compiler.parse(source);
  return compiler.getBettiNumber();
}

/**
 * Compose two topology sources by concatenating them.
 * The composition connects exported nodes from A to imported nodes in B.
 * For simplicity: concatenation merges the graphs (shared node IDs connect).
 */
function composeTopologies(sources: string[]): string {
  return sources.join('\n\n// ── composition boundary ──\n\n');
}

// ── Brunnian Detection ───────────────────────────────────────

/**
 * Detect Brunnian coupling across a set of topology modules.
 *
 * Algorithm:
 * 1. Compute beta1 for each pairwise composition
 * 2. Compute beta1 for the full n-wise composition
 * 3. If full > sum(pairwise), emergent coupling exists
 */
export function detectBrunnianCoupling(
  modules: TopologyModule[]
): BrunnianResult {
  try {
    if (modules.length < 3) {
      return {
        modules: modules.map((m) => m.id),
        composedBeta1: 0,
        pairwiseBeta1Sum: 0,
        emergentStrength: 0,
        isBrunnian: false,
      };
    }

    // Compute pairwise beta1 values
    let pairwiseSum = 0;
    for (let i = 0; i < modules.length; i++) {
      for (let j = i + 1; j < modules.length; j++) {
        const pairSource = composeTopologies([
          modules[i].source,
          modules[j].source,
        ]);
        const pairBeta1 = computeBeta1(pairSource);
        pairwiseSum += pairBeta1;
      }
    }

    // Compute full n-wise composition beta1
    const fullSource = composeTopologies(modules.map((m) => m.source));
    const composedBeta1 = computeBeta1(fullSource);

    const emergentStrength = composedBeta1 - pairwiseSum;

    return {
      modules: modules.map((m) => m.id),
      composedBeta1,
      pairwiseBeta1Sum: pairwiseSum,
      emergentStrength,
      isBrunnian: emergentStrength > 0,
    };

  } catch (error) {
    throw new Error(`[detectBrunnianCoupling] ${error instanceof Error ? error.message : String(error)}`);
  }
}

// ── Test Generation ──────────────────────────────────────────

/**
 * Generate test cases from Brunnian detection results.
 * Each test exercises the emergent coupling that pairwise tests miss.
 */
export function generateBrunnianTests(
  modules: TopologyModule[]
): BrunnianTestCase[] {
  try {
    const tests: BrunnianTestCase[] = [];

    // For each triple (or n-tuple), check for Brunnian coupling
    const n = modules.length;

    // Generate all 3-wise combinations
    for (let i = 0; i < n; i++) {
      for (let j = i + 1; j < n; j++) {
        for (let k = j + 1; k < n; k++) {
          const triple = [modules[i], modules[j], modules[k]];
          const result = detectBrunnianCoupling(triple);

          if (result.isBrunnian) {
            tests.push({
              description: `Brunnian coupling: ${result.modules.join(
                ' + '
              )} (emergent beta1 = ${result.emergentStrength})`,
              composition: result.modules,
              expectedFailure: classifyFailureMode(result),
              emergentBeta1: result.emergentStrength,
            });
          }
        }
      }
    }

    // Also check 4-wise if n >= 4 (higher-order Brunnian links)
    if (n >= 4) {
      for (let i = 0; i < n; i++) {
        for (let j = i + 1; j < n; j++) {
          for (let k = j + 1; k < n; k++) {
            for (let l = k + 1; l < n; l++) {
              const quad = [modules[i], modules[j], modules[k], modules[l]];
              const result = detectBrunnianCoupling(quad);

              if (result.isBrunnian) {
                tests.push({
                  description: `4-way Brunnian: ${result.modules.join(
                    ' + '
                  )} (emergent beta1 = ${result.emergentStrength})`,
                  composition: result.modules,
                  expectedFailure: classifyFailureMode(result),
                  emergentBeta1: result.emergentStrength,
                });
              }
            }
          }
        }
      }
    }

    return tests;

  } catch (error) {
    throw new Error(`[generateBrunnianTests] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Classify the expected failure mode based on the emergent coupling pattern.
 */
function classifyFailureMode(
  result: BrunnianResult
): BrunnianTestCase['expectedFailure'] {
  try {
    // High emergent strength with many modules = likely deadlock (circular dependency)
    if (result.emergentStrength >= 3 && result.modules.length >= 3)
      return 'deadlock';
    // Moderate strength = likely race condition (braided execution)
    if (result.emergentStrength >= 2) return 'race';
    // Low strength = likely resource starvation
    if (result.emergentStrength >= 1) return 'starvation';
    return 'unknown';

  } catch (error) {
    throw new Error(`[classifyFailureMode] ${error instanceof Error ? error.message : String(error)}`);
  }
}

// ── Void Boundary Feeding ────────────────────────────────────

/**
 * Convert Brunnian detection results into void boundary rejections.
 * These feed buleyean-rl training: the model learns which compositions
 * produce emergent failures.
 */
export function brunnianToVoidRejections(
  results: BrunnianResult[]
): VoidRejection[] {
  return results
    .filter((r) => r.isBrunnian)
    .map((r) => ({
      source: 'brunnian-detector' as const,
      composition: r.modules,
      reason: `Emergent coupling: beta1(composed)=${r.composedBeta1} > sum(pairwise)=${r.pairwiseBeta1Sum}. Strength=${r.emergentStrength}.`,
      weight: r.emergentStrength,
      timestamp: new Date().toISOString(),
    }));
}

// ── Scanner Rule Integration ─────────────────────────────────

export interface ScannerFinding {
  rule: string;
  level: 'error' | 'warning' | 'hint';
  message: string;
  filePath: string;
  line: number;
}

/**
 * Generate scanner findings from Brunnian analysis.
 * These appear in the polyglot scanner as BRUNNIAN_COUPLING warnings.
 */
export function brunnianToFindings(
  results: BrunnianResult[],
  filePath: string
): ScannerFinding[] {
  return results
    .filter((r) => r.isBrunnian)
    .map((r) => ({
      rule: 'BRUNNIAN_COUPLING',
      level:
        r.emergentStrength >= 3 ? ('error' as const) : ('warning' as const),
      message:
        `Emergent coupling detected: ${r.modules.join(
          ' + '
        )} compose to beta1=${r.composedBeta1} (pairwise sum=${
          r.pairwiseBeta1Sum
        }). ` +
        `This ${
          r.emergentStrength >= 3 ? 'will' : 'may'
        } cause failures invisible to unit and pairwise integration tests. ` +
        `Test the full ${r.modules.length}-way composition.`,
      filePath,
      line: 1,
    }));
}

// ── .test.gg Topology Generation ─────────────────────────────

/**
 * Generate a .test.gg topology that exercises Brunnian coupling.
 * The test topology forks all modules in parallel, races their
 * outputs, and checks if the fold produces a consistent result.
 */
export function generateTestTopology(test: BrunnianTestCase): string {
  const moduleNodes = test.composition.map(
    (id, i) => `(mod_${i}: Module { id: '${id}' })`
  );
  const forkTargets = test.composition.map((_, i) => `mod_${i}`).join(' | ');
  const raceTargets = test.composition.map((_, i) => `mod_${i}`).join(' | ');

  return [
    `// Brunnian coupling test: ${test.description}`,
    `// Expected failure: ${test.expectedFailure}`,
    `// Emergent beta1: ${test.emergentBeta1}`,
    '',
    `(input: TestInput { composition: '${test.composition.join(',')}' })`,
    ...moduleNodes,
    `(result: TestResult)`,
    `(void: VoidBoundary)`,
    '',
    `// Fork: launch all modules in parallel`,
    `(input)-[:FORK]->(${forkTargets})`,
    '',
    `// Race: wait for all to complete (the Brunnian test)`,
    `(${raceTargets})-[:RACE]->(result)`,
    '',
    `// Vent: capture failures as void boundary rejections`,
    `(result)-[:VENT]->(void)`,
  ].join('\n');
}
