// @a0n/.ignored_gnosis/src/buleyean-kernel-shim.ts
/**
 * Local SSOT mirror of `@a0n/buleyean-kernel` for gnosis runtime paths.
 *
 * Launchd/tsx often resolve gnosis sources without the workspace linker putting
 * `packages/buleyean-kernel` on NODE_PATH. The formula is one line — keep it
 * here in sync with `packages/buleyean-kernel/src/index.ts`.
 */

/** Buleyean weight `w(R, v) = R - min(v, R) + 1`. */
export function buleyeanWeight(rounds: number, rejections: number): number {
  return rounds - Math.min(rejections, rounds) + 1;
}

export const deceptaconWeight = buleyeanWeight;

/**
 * Handles the gnosis bule Deficit workflow.
 */
export function buleDeficit(choices: number, rejections: number): number {
  return Math.max(0, choices - 1 - rejections);
}

/**
 * Handles the gnosis deficit At Round workflow.
 */
export function deficitAtRound(deficit: number, k: number): number {
  return Math.max(0, deficit - k);
}

/**
 * Returns whether is Buleyean Positive is true.
 */
export function isBuleyeanPositive(
  rejections: readonly number[],
  rounds: number
): boolean {
  for (const v of rejections) {
    if (buleyeanWeight(rounds, v) <= 1) return false;
  }
  return true;
}
