// @a0n/.ignored_gnosis/src/causal-diamond.ts
/**
 * Causal Diamond — the intersection of past and future light cones.
 *
 * An event E lies inside the causal diamond of an observer with birth B
 * and death D if and only if:
 *   - E is in the FUTURE light cone of B (B could have caused E)
 *   - E is in the PAST light cone of D (E could have caused D)
 *
 * In 1+1D Penrose coordinates the diamond boundary is literally a
 * diamond shape formed by four 45-degree null lines emanating from
 * the birth and death events.
 *
 * @packageDocumentation
 */

import {
  type SpacetimeEvent,
  lightConeRegion,
  penroseCoordinates,
  type PenroseCoordinate,
} from './causal-topology';
import { LIGHT_PICOLORENZO_LY } from './math';

// ─── Types ─────────────────────────────────────────────────────────

export interface CausalDiamondBody {
  readonly event: SpacetimeEvent;
  readonly penrose: PenroseCoordinate;
  readonly insideDiamond: boolean;
}

export interface CausalDiamond {
  readonly source: 'GnosisCausalDiamond';
  readonly observer: SpacetimeEvent;
  readonly agePLo: number;
  readonly birthEvent: SpacetimeEvent;
  readonly deathEvent: SpacetimeEvent;
  readonly boundaryPenrose: readonly PenroseCoordinate[];
  readonly bodies: readonly CausalDiamondBody[];
  readonly insideCount: number;
  readonly outsideCount: number;
  readonly insideFraction: number;
}

// ─── Core Predicate ────────────────────────────────────────────────

/**
 * Check whether a test event lies inside the causal diamond defined
 * by a birth event and a death event.
 *
 * The diamond is the intersection of the future light cone of birth
 * and the past light cone of death. An event at either endpoint
 * ('here') also counts as inside.
 */
export function isInsideCausalDiamond(
  birthEvent: SpacetimeEvent,
  deathEvent: SpacetimeEvent,
  testEvent: SpacetimeEvent
): boolean {
  const fromBirth = lightConeRegion(birthEvent, testEvent);
  const fromDeath = lightConeRegion(deathEvent, testEvent);

  const inFutureOfBirth = fromBirth === 'future' || fromBirth === 'here';
  const inPastOfDeath = fromDeath === 'past' || fromDeath === 'here';

  return inFutureOfBirth && inPastOfDeath;
}

// ─── Boundary Generation ───────────────────────────────────────────

/**
 * Generate boundary points of the causal diamond in Penrose coordinates.
 *
 * The diamond has four vertices in spacetime (relative to origin at the
 * midpoint between birth and death):
 *   - Birth: bottom vertex (t = birthPLo, r = 0)
 *   - Death: top vertex (t = deathPLo, r = 0)
 *   - Right tip: (t = midpointPLo, r = halfAgeLpLo)
 *   - Left tip: (t = midpointPLo, r = -halfAgeLpLo)
 *
 * The boundary traces four null (45-degree) segments:
 *   birth -> right tip -> death -> left tip -> birth
 *
 * Each boundary point lies on a light cone from either birth or death.
 */
export function causalDiamondBoundary(
  birthEvent: SpacetimeEvent,
  deathEvent: SpacetimeEvent,
  origin: SpacetimeEvent,
  boundaryPoints: number = 40
): readonly PenroseCoordinate[] {
  const agePLo = deathEvent.pLo - birthEvent.pLo;
  const halfAgePLo = agePLo / 2;
  // Half-age in light-years (spatial reach at the widest point)
  const halfAgeLy = halfAgePLo * LIGHT_PICOLORENZO_LY;
  const midpointPLo = birthEvent.pLo + halfAgePLo;

  // Points per segment (4 segments around the diamond)
  const perSegment = Math.max(1, Math.floor(boundaryPoints / 4));
  const spacetimePoints: SpacetimeEvent[] = [];

  // Segment 1: birth -> right tip
  for (let i = 0; i < perSegment; i++) {
    const frac = i / perSegment;
    spacetimePoints.push({
      id: `boundary-${spacetimePoints.length}`,
      label: 'boundary',
      x: birthEvent.x + frac * halfAgeLy,
      y: birthEvent.y,
      z: birthEvent.z,
      pLo: birthEvent.pLo + frac * halfAgePLo,
    });
  }

  // Segment 2: right tip -> death
  for (let i = 0; i < perSegment; i++) {
    const frac = i / perSegment;
    spacetimePoints.push({
      id: `boundary-${spacetimePoints.length}`,
      label: 'boundary',
      x: birthEvent.x + halfAgeLy * (1 - frac),
      y: birthEvent.y,
      z: birthEvent.z,
      pLo: midpointPLo + frac * halfAgePLo,
    });
  }

  // Segment 3: death -> left tip
  for (let i = 0; i < perSegment; i++) {
    const frac = i / perSegment;
    spacetimePoints.push({
      id: `boundary-${spacetimePoints.length}`,
      label: 'boundary',
      x: birthEvent.x - frac * halfAgeLy,
      y: birthEvent.y,
      z: birthEvent.z,
      pLo: deathEvent.pLo - frac * halfAgePLo,
    });
  }

  // Segment 4: left tip -> birth
  for (let i = 0; i < perSegment; i++) {
    const frac = i / perSegment;
    spacetimePoints.push({
      id: `boundary-${spacetimePoints.length}`,
      label: 'boundary',
      x: birthEvent.x - halfAgeLy * (1 - frac),
      y: birthEvent.y,
      z: birthEvent.z,
      pLo: birthEvent.pLo + halfAgePLo * (1 - frac),
    });
  }

  return penroseCoordinates(spacetimePoints, origin);
}

// ─── Main Entry Point ──────────────────────────────────────────────

/**
 * Construct a causal diamond for an observer.
 *
 * Birth = observer shifted back in time by agePLo (same spatial position,
 * pLo - agePLo). Death = observer at their current pLo.
 *
 * Classifies all supplied events as inside or outside the diamond,
 * computes the Penrose boundary, and returns the full diamond packet.
 */
export function causalDiamond(
  observer: SpacetimeEvent,
  agePLo: number,
  events: readonly SpacetimeEvent[]
): CausalDiamond {
  const birthEvent: SpacetimeEvent = {
    id: `${observer.id}-birth`,
    label: `${observer.label} birth`,
    x: observer.x,
    y: observer.y,
    z: observer.z,
    pLo: observer.pLo - agePLo,
  };

  const deathEvent: SpacetimeEvent = {
    id: `${observer.id}-death`,
    label: `${observer.label} death`,
    x: observer.x,
    y: observer.y,
    z: observer.z,
    pLo: observer.pLo,
  };

  // Use the midpoint between birth and death as Penrose origin
  const midpoint: SpacetimeEvent = {
    id: 'diamond-origin',
    label: 'diamond origin',
    x: observer.x,
    y: observer.y,
    z: observer.z,
    pLo: observer.pLo - agePLo / 2,
  };

  const penroseAll = penroseCoordinates(events, midpoint);
  let insideCount = 0;

  const bodies: CausalDiamondBody[] = events.map((event, i) => {
    const inside = isInsideCausalDiamond(birthEvent, deathEvent, event);
    if (inside) insideCount++;
    return {
      event,
      penrose: penroseAll[i]!,
      insideDiamond: inside,
    };
  });

  const outsideCount = events.length - insideCount;
  const boundaryPenrose = causalDiamondBoundary(birthEvent, deathEvent, midpoint);

  return {
    source: 'GnosisCausalDiamond',
    observer,
    agePLo,
    birthEvent,
    deathEvent,
    boundaryPenrose,
    bodies,
    insideCount,
    outsideCount,
    insideFraction: events.length > 0 ? insideCount / events.length : 0,
  };
}
