// @a0n/.ignored_gnosis/src/causal-topology.ts
/**
 * Causal Topology — relativistic spacetime structure between events.
 *
 * Every pair of events in spacetime has a single invariant number that
 * all observers agree on: the spacetime interval. This module computes
 * that interval and builds a directed graph (topology) whose edges
 * encode causal relationships:
 *
 *   PROCESS = timelike (one event can cause the other)
 *   RACE    = lightlike (a photon connects them -- the boundary)
 *   VENT    = spacelike (no causal connection, different "nows")
 *
 * The light-picolorenzo (lpLo) is the natural unit: the distance
 * light travels in one picolorenzo (≈ 0.00858 ly ≈ 1.816 AU).
 *
 * @packageDocumentation
 */

import {
  LIGHT_PICOLORENZO_LY,
  lightTravelTimePLo,
  formatGnosticTime,
} from './math';

// ─── Spacetime Event ────────────────────────────────────────────────

/**
 * A point in spacetime. Position in light-years (any consistent frame)
 * and time in picolorenzos since Unix epoch.
 */
export interface SpacetimeEvent {
  readonly id: string;
  readonly label: string;
  /** Cartesian position in light-years. */
  readonly x: number;
  readonly y: number;
  readonly z: number;
  /** Time in picolorenzos since Unix epoch. */
  readonly pLo: number;
}

// ─── Interval Classification ────────────────────────────────────────

/**
 * Spacetime interval classification.
 *
 * Using the (-,+,+,+) metric signature (particle physics convention):
 *   ds² = -c²dt² + dx² + dy² + dz²
 *
 * - Timelike (ds² < 0): causal connection possible. The interval
 *   represents proper time experienced by a traveler.
 * - Lightlike (ds² = 0): on the light cone boundary. Only photons
 *   travel this path.
 * - Spacelike (ds² > 0): no causal connection. The events exist in
 *   different "nows" for different observers.
 */
export type IntervalClass = 'timelike' | 'lightlike' | 'spacelike';

/**
 * The gnosis edge type that corresponds to each interval class.
 * This maps relativistic causality directly to the topology language.
 */
export type CausalEdgeType = 'PROCESS' | 'RACE' | 'VENT';

/**
 * Full spacetime interval between two events.
 */
export interface SpacetimeInterval {
  /** The two events. */
  readonly from: SpacetimeEvent;
  readonly to: SpacetimeEvent;

  /** Spatial separation in light-years. */
  readonly spatialDistanceLy: number;
  /** Temporal separation in picolorenzos. */
  readonly temporalSeparationPLo: number;

  /**
   * The invariant interval ds².
   * Negative = timelike, zero = lightlike, positive = spacelike.
   * Units: (light-picolorenzos)².
   *
   * ds² = -(c·dt)² + dx² + dy² + dz²
   * In lpLo units: ds² = -dt² + dr² (since c = 1 lpLo/pLo)
   */
  readonly dsSquared: number;

  /** Classification. */
  readonly classification: IntervalClass;

  /** Corresponding gnosis edge type. */
  readonly edgeType: CausalEdgeType;

  /**
   * For timelike intervals: proper time experienced by a traveler
   * taking the straight (geodesic) path, in picolorenzos.
   * τ = √(-ds²). Undefined for spacelike intervals.
   */
  readonly properTimePLo: number | null;

  /**
   * For spacelike intervals: proper distance measured by an observer
   * for whom the events are simultaneous, in light-years.
   * d = √(ds²) × LIGHT_PICOLORENZO_LY. Undefined for timelike intervals.
   */
  readonly properDistanceLy: number | null;

  /**
   * Light travel time across the spatial distance, in picolorenzos.
   * This is how long a photon takes to cross the gap.
   */
  readonly lightCrossingPLo: number;
}

// ─── Light Cone ─────────────────────────────────────────────────────

/** Causal relationship of an event to a reference event's light cone. */
export type LightConeRegion =
  | 'past'       // in the past light cone (could have caused the reference)
  | 'future'     // in the future light cone (reference could cause it)
  | 'elsewhere'  // spacelike separated (no causal connection)
  | 'here';      // same event (or within resolution threshold)

/**
 * A body's membership in another body's light cone.
 */
export interface LightConeMembership {
  readonly event: SpacetimeEvent;
  readonly region: LightConeRegion;
  readonly interval: SpacetimeInterval;
}

/**
 * Full light cone structure from a reference event.
 */
export interface LightCone {
  readonly reference: SpacetimeEvent;
  readonly members: readonly LightConeMembership[];
  readonly pastCount: number;
  readonly futureCount: number;
  readonly elsewhereCount: number;
}

// ─── Lookback ───────────────────────────────────────────────────────

/**
 * Lookback correction for a distant body as seen by an observer.
 *
 * When you look at a body N light-years away, you see it as it was
 * N years ago. The "observed pLo" is the time the light left the body;
 * the "observer pLo" is when the observation happens.
 */
export interface LookbackCorrection {
  readonly body: SpacetimeEvent;
  /** When the observer sees this body (their "now"). */
  readonly observerPLo: number;
  /** When the light actually left the body. */
  readonly emittedPLo: number;
  /** How far back in time the observer is seeing, in picolorenzos. */
  readonly lookbackPLo: number;
  /** How far back in years. */
  readonly lookbackYears: number;
  /** Gnostic time label for the lookback. */
  readonly lookbackGnostic: string;
  /** Spatial distance in light-years. */
  readonly distanceLy: number;
}

// ─── Causal Graph ───────────────────────────────────────────────────

/**
 * An edge in the causal graph.
 */
export interface CausalEdge {
  readonly sourceId: string;
  readonly targetId: string;
  readonly edgeType: CausalEdgeType;
  readonly interval: SpacetimeInterval;
}

/**
 * The causal topology packet: a directed graph over spacetime events
 * where edges encode causal structure.
 */
export interface CausalTopologyPacket {
  readonly source: 'GnosisCausalTopology';
  readonly events: readonly SpacetimeEvent[];
  readonly edges: readonly CausalEdge[];
  /** All timelike edges (causal chains). */
  readonly timelikeEdges: readonly CausalEdge[];
  /** All lightlike edges (photon paths). */
  readonly lightlikeEdges: readonly CausalEdge[];
  /** All spacelike edges (no causal connection). */
  readonly spacelikeEdges: readonly CausalEdge[];
  /** Summary statistics. */
  readonly summary: CausalSummary;
}

export interface CausalSummary {
  readonly eventCount: number;
  readonly edgeCount: number;
  readonly timelikeCount: number;
  readonly lightlikeCount: number;
  readonly spacelikeCount: number;
  /** Fraction of pairs that are causally connected (timelike + lightlike). */
  readonly causalConnectivity: number;
  /** Diameter: maximum light-crossing time across all pairs, in picolorenzos. */
  readonly diameterPLo: number;
  /** The most causally isolated event (most spacelike edges). */
  readonly mostIsolatedEventId: string;
  /** The most causally connected event (most timelike/lightlike edges). */
  readonly mostConnectedEventId: string;
}

// ─── Spacetime Interval Computation ─────────────────────────────────

/** Lightlike threshold: intervals with |ds²| below this are classified as lightlike. */
const LIGHTLIKE_EPSILON = 1e-6;

/**
 * Compute the spacetime interval between two events.
 *
 * Uses the (-,+,+,+) metric signature with natural units where
 * c = 1 light-picolorenzo per picolorenzo.
 */
export function spacetimeInterval(
  from: SpacetimeEvent,
  to: SpacetimeEvent
): SpacetimeInterval {
  const dx = to.x - from.x;
  const dy = to.y - from.y;
  const dz = to.z - from.z;
  const spatialDistanceLy = Math.sqrt(dx * dx + dy * dy + dz * dz);

  const temporalSeparationPLo = to.pLo - from.pLo;

  // Convert to natural units (light-picolorenzos for space, picolorenzos for time)
  const drLpLo = spatialDistanceLy / LIGHT_PICOLORENZO_LY; // spatial distance in lpLo
  const dtPLo = temporalSeparationPLo;                     // time already in pLo

  // ds² = -dt² + dr² (c = 1 in natural units)
  const dsSquared = -(dtPLo * dtPLo) + drLpLo * drLpLo;

  let classification: IntervalClass;
  if (Math.abs(dsSquared) < LIGHTLIKE_EPSILON) {
    classification = 'lightlike';
  } else if (dsSquared < 0) {
    classification = 'timelike';
  } else {
    classification = 'spacelike';
  }

  const edgeType: CausalEdgeType =
    classification === 'timelike' ? 'PROCESS' :
    classification === 'lightlike' ? 'RACE' : 'VENT';

  // Proper time (timelike only): τ = √(-ds²)
  const properTimePLo = classification === 'timelike'
    ? Math.sqrt(-dsSquared) : null;

  // Proper distance (spacelike only): d = √(ds²) converted to light-years
  const properDistanceLy = classification === 'spacelike'
    ? Math.sqrt(dsSquared) * LIGHT_PICOLORENZO_LY : null;

  const lightCrossingPLo = lightTravelTimePLo(spatialDistanceLy);

  return {
    from,
    to,
    spatialDistanceLy,
    temporalSeparationPLo,
    dsSquared,
    classification,
    edgeType,
    properTimePLo,
    properDistanceLy,
    lightCrossingPLo,
  };
}

// ─── Light Cone ─────────────────────────────────────────────────────

/** Threshold for considering two events at the same location. */
const HERE_EPSILON_LY = 1e-9;
const HERE_EPSILON_PLO = 1e-9;

/**
 * Classify which region of the reference event's light cone
 * a given event falls in.
 */
export function lightConeRegion(
  reference: SpacetimeEvent,
  event: SpacetimeEvent
): LightConeRegion {
  const dx = event.x - reference.x;
  const dy = event.y - reference.y;
  const dz = event.z - reference.z;
  const dist = Math.sqrt(dx * dx + dy * dy + dz * dz);
  const dt = event.pLo - reference.pLo;

  // Same event check
  if (dist < HERE_EPSILON_LY && Math.abs(dt) < HERE_EPSILON_PLO) {
    return 'here';
  }

  const lightDistLpLo = dist / LIGHT_PICOLORENZO_LY;

  if (Math.abs(dt) >= lightDistLpLo - LIGHTLIKE_EPSILON) {
    // Inside or on the light cone
    return dt > 0 ? 'future' : 'past';
  }

  return 'elsewhere';
}

/**
 * Compute the full light cone structure from a reference event.
 */
export function lightCone(
  reference: SpacetimeEvent,
  events: readonly SpacetimeEvent[]
): LightCone {
  let pastCount = 0;
  let futureCount = 0;
  let elsewhereCount = 0;

  const members: LightConeMembership[] = [];

  for (const event of events) {
    if (event.id === reference.id) continue;
    const region = lightConeRegion(reference, event);
    const interval = spacetimeInterval(reference, event);

    if (region === 'past') pastCount++;
    else if (region === 'future') futureCount++;
    else if (region === 'elsewhere') elsewhereCount++;

    members.push({ event, region, interval });
  }

  return {
    reference,
    members,
    pastCount,
    futureCount,
    elsewhereCount,
  };
}

// ─── Lookback ───────────────────────────────────────────────────────

/**
 * Compute the lookback correction for a distant body.
 *
 * When you observe a body at distance d, you see it as it was
 * d/c time ago. This function computes that correction.
 */
export function lookbackCorrection(
  body: SpacetimeEvent,
  observerPosition: { x: number; y: number; z: number },
  observerPLo: number
): LookbackCorrection {
  const dx = body.x - observerPosition.x;
  const dy = body.y - observerPosition.y;
  const dz = body.z - observerPosition.z;
  const distanceLy = Math.sqrt(dx * dx + dy * dy + dz * dz);

  const lookbackPLo = lightTravelTimePLo(distanceLy);
  const emittedPLo = observerPLo - lookbackPLo;
  const lookbackYears = distanceLy; // 1 light-year of distance = 1 year of lookback

  return {
    body,
    observerPLo,
    emittedPLo,
    lookbackPLo,
    lookbackYears,
    lookbackGnostic: formatGnosticTime(lookbackPLo),
    distanceLy,
  };
}

/**
 * Apply lookback correction to a list of bodies as seen from an observer.
 * Returns each body with its "emitted time" -- when the light you're
 * seeing actually left.
 */
export function lookbackCorrectedView(
  bodies: readonly SpacetimeEvent[],
  observerPosition: { x: number; y: number; z: number },
  observerPLo: number
): readonly LookbackCorrection[] {
  return bodies.map((body) =>
    lookbackCorrection(body, observerPosition, observerPLo)
  );
}

// ─── Causal Topology Construction ───────────────────────────────────

/**
 * Build a causal topology: a directed graph over spacetime events
 * where every pair of events gets an edge classified by its spacetime
 * interval.
 *
 * The resulting graph is the causal structure of the universe as seen
 * from these events. Timelike edges (PROCESS) are causal chains.
 * Lightlike edges (RACE) are photon paths. Spacelike edges (VENT) are
 * acausal separations.
 *
 * For N events this produces N*(N-1)/2 edges (the complete graph).
 * The direction on timelike/lightlike edges follows the arrow of time
 * (from earlier pLo to later pLo). Spacelike edges are undirected
 * (arbitrarily ordered by id).
 */
export function createCausalTopology(
  events: readonly SpacetimeEvent[]
): CausalTopologyPacket {
  const edges: CausalEdge[] = [];
  const timelikeEdges: CausalEdge[] = [];
  const lightlikeEdges: CausalEdge[] = [];
  const spacelikeEdges: CausalEdge[] = [];

  // Track per-event connectivity for summary
  const spacelikeCount = new Map<string, number>();
  const causalCount = new Map<string, number>();
  for (const e of events) {
    spacelikeCount.set(e.id, 0);
    causalCount.set(e.id, 0);
  }

  let maxLightCrossing = 0;

  for (let i = 0; i < events.length; i++) {
    for (let j = i + 1; j < events.length; j++) {
      const a = events[i]!;
      const b = events[j]!;
      const interval = spacetimeInterval(a, b);

      // Direction: earlier event is source
      const [source, target] = a.pLo <= b.pLo ? [a, b] : [b, a];

      const edge: CausalEdge = {
        sourceId: source.id,
        targetId: target.id,
        edgeType: interval.edgeType,
        interval,
      };

      edges.push(edge);

      if (interval.classification === 'timelike') {
        timelikeEdges.push(edge);
        causalCount.set(source.id, (causalCount.get(source.id) ?? 0) + 1);
        causalCount.set(target.id, (causalCount.get(target.id) ?? 0) + 1);
      } else if (interval.classification === 'lightlike') {
        lightlikeEdges.push(edge);
        causalCount.set(source.id, (causalCount.get(source.id) ?? 0) + 1);
        causalCount.set(target.id, (causalCount.get(target.id) ?? 0) + 1);
      } else {
        spacelikeEdges.push(edge);
        spacelikeCount.set(source.id, (spacelikeCount.get(source.id) ?? 0) + 1);
        spacelikeCount.set(target.id, (spacelikeCount.get(target.id) ?? 0) + 1);
      }

      if (interval.lightCrossingPLo > maxLightCrossing) {
        maxLightCrossing = interval.lightCrossingPLo;
      }
    }
  }

  // Find most isolated and most connected
  let mostIsolatedId = events[0]?.id ?? '';
  let maxSpacelike = 0;
  let mostConnectedId = events[0]?.id ?? '';
  let maxCausal = 0;

  for (const e of events) {
    const sc = spacelikeCount.get(e.id) ?? 0;
    const cc = causalCount.get(e.id) ?? 0;
    if (sc > maxSpacelike) { maxSpacelike = sc; mostIsolatedId = e.id; }
    if (cc > maxCausal) { maxCausal = cc; mostConnectedId = e.id; }
  }

  const totalPairs = edges.length;
  const causalPairs = timelikeEdges.length + lightlikeEdges.length;

  return {
    source: 'GnosisCausalTopology',
    events,
    edges,
    timelikeEdges,
    lightlikeEdges,
    spacelikeEdges,
    summary: {
      eventCount: events.length,
      edgeCount: edges.length,
      timelikeCount: timelikeEdges.length,
      lightlikeCount: lightlikeEdges.length,
      spacelikeCount: spacelikeEdges.length,
      causalConnectivity: totalPairs > 0 ? causalPairs / totalPairs : 0,
      diameterPLo: maxLightCrossing,
      mostIsolatedEventId: mostIsolatedId,
      mostConnectedEventId: mostConnectedId,
    },
  };
}

// ─── Visualization Helpers ──────────────────────────────────────────

/**
 * Color encoding for interval classes.
 * Timelike = warm (causal heat). Spacelike = cool (isolation).
 * Lightlike = gold (the boundary, the horizon).
 */
export const CAUSAL_COLORS = {
  timelike:  { r: 255, g:  89, b:  94, a: 1.0 },  // warm coral
  lightlike: { r: 255, g: 202, b:  58, a: 1.0 },  // gold
  spacelike: { r:  86, g: 207, b: 225, a: 1.0 },  // cool cyan
  here:      { r: 255, g: 255, b: 255, a: 1.0 },  // white
} as const;

/**
 * Map an interval class to a CSS color.
 */
export function causalColor(classification: IntervalClass | 'here'): string {
  const c = CAUSAL_COLORS[classification];
  return `rgba(${c.r},${c.g},${c.b},${c.a})`;
}

/**
 * Compute the "causal temperature" of an event: the fraction of its
 * connections that are causal (timelike + lightlike) vs total.
 * 1.0 = fully connected to everything (hot). 0.0 = fully isolated (cold).
 */
export function causalTemperature(
  eventId: string,
  topology: CausalTopologyPacket
): number {
  let causal = 0;
  let total = 0;
  for (const edge of topology.edges) {
    if (edge.sourceId === eventId || edge.targetId === eventId) {
      total++;
      if (edge.edgeType !== 'VENT') causal++;
    }
  }
  return total > 0 ? causal / total : 0;
}

/**
 * Compute the "horizon depth" of an event: how many picolorenzos
 * of lookback time separate it from the most distant event in
 * its past light cone.
 */
export function horizonDepth(
  eventId: string,
  topology: CausalTopologyPacket
): number {
  let maxLookback = 0;
  for (const edge of topology.timelikeEdges) {
    if (edge.targetId === eventId) {
      const lookback = Math.abs(edge.interval.temporalSeparationPLo);
      if (lookback > maxLookback) maxLookback = lookback;
    }
  }
  for (const edge of topology.lightlikeEdges) {
    if (edge.targetId === eventId) {
      const lookback = Math.abs(edge.interval.temporalSeparationPLo);
      if (lookback > maxLookback) maxLookback = lookback;
    }
  }
  return maxLookback;
}

// ─── Penrose Diagram Coordinates ────────────────────────────────────

/**
 * Penrose diagram coordinates for a spacetime event relative to an origin.
 *
 * The Penrose (conformal) diagram compresses infinite spacetime into
 * a finite diamond. It uses null coordinates:
 *   u = t - r  (retarded time)
 *   v = t + r  (advanced time)
 *
 * Then applies arctan compression so infinity maps to π/2:
 *   U = arctan(u) / (π/2)  →  [-1, 1]
 *   V = arctan(v) / (π/2)  →  [-1, 1]
 *
 * Finally, rotate to standard Penrose orientation:
 *   penroseTime  = (V + U) / 2   (vertical axis)
 *   penroseSpace = (V - U) / 2   (horizontal axis)
 *
 * Light cones are always 45-degree lines in this diagram.
 */
export interface PenroseCoordinate {
  readonly eventId: string;
  /** Penrose time coordinate [-1, 1]. */
  readonly penroseTime: number;
  /** Penrose space coordinate [0, 1] (radial, always positive). */
  readonly penroseSpace: number;
  /** Null coordinate u = t - r. */
  readonly u: number;
  /** Null coordinate v = t + r. */
  readonly v: number;
}

/**
 * Compute Penrose diagram coordinates for events relative to an origin.
 * All coordinates are in natural units (picolorenzos for time,
 * light-picolorenzos for space).
 */
export function penroseCoordinates(
  events: readonly SpacetimeEvent[],
  origin: SpacetimeEvent
): readonly PenroseCoordinate[] {
  const HALF_PI = Math.PI / 2;

  return events.map((event) => {
    const dx = event.x - origin.x;
    const dy = event.y - origin.y;
    const dz = event.z - origin.z;
    const rLy = Math.sqrt(dx * dx + dy * dy + dz * dz);
    const rLpLo = rLy / LIGHT_PICOLORENZO_LY;
    const tPLo = event.pLo - origin.pLo;

    const u = tPLo - rLpLo;
    const v = tPLo + rLpLo;

    const U = Math.atan(u) / HALF_PI;
    const V = Math.atan(v) / HALF_PI;

    return {
      eventId: event.id,
      penroseTime: (V + U) / 2,
      penroseSpace: (V - U) / 2,
      u,
      v,
    };
  });
}
