// @a0n/.ignored_gnosis/src/certified-allocation-kernel.ts
export const CERTIFIED_ALLOCATION_KERNEL_THEOREM_IDS = [
  'THM-CERTIFIED-ALLOCATION-CONFIDENCE-FLOOR',
  'THM-CERTIFIED-ALLOCATION-CALIBRATION-DEFICIT',
  'THM-CERTIFIED-ALLOCATION-CALIBRATION-CONSERVATION',
  'THM-CERTIFIED-ALLOCATION-SHARP-ADMISSION',
  'THM-CERTIFIED-ALLOCATION-FINITE-TERMINATION',
  'THM-CERTIFIED-ALLOCATION-DEADLINE-FAIL-CLOSED',
  'THM-CERTIFIED-ALLOCATION-LOOP-PROGRESS-OR-EXPLICIT-FAILURE',
] as const;

export type CertifiedAllocationKernelTheoremId =
  (typeof CERTIFIED_ALLOCATION_KERNEL_THEOREM_IDS)[number];

export type AdmittedObservationZone =
  | 'top-left'
  | 'top-center'
  | 'top-right'
  | 'center-left'
  | 'center'
  | 'center-right'
  | 'bottom-left'
  | 'bottom-center'
  | 'bottom-right';

export type ObservationZone = AdmittedObservationZone | 'unavailable';

export interface NormalizedMilliPoint {
  readonly x: number;
  readonly y: number;
}

export interface CertifiedObservationFrame {
  readonly gazePoint: NormalizedMilliPoint | null;
  readonly zone: ObservationZone;
  readonly confidenceScore: number;
  readonly confidenceFloor: number;
}

export interface CertifiedCalibrationPlan {
  readonly pointIds: readonly string[];
  readonly samplesPerPoint: number;
  readonly maxAttemptsPerPoint: number;
}

export interface CertifiedCalibrationPointLedger {
  readonly pointId: string;
  readonly accepted: number;
  readonly rejected: number;
  readonly observed: number;
}

export interface CertifiedLoopBudget {
  readonly deadlineMs: number;
  readonly elapsedMs: number;
}

export type CertifiedSafeStateReason =
  | 'deadline-exceeded'
  | 'not-calibrated'
  | 'observation-unavailable'
  | 'no-candidates';

export type CertifiedAllocationDecision =
  | {
      readonly kind: 'allocate';
      readonly candidateId: number;
    }
  | {
      readonly kind: 'safe-state';
      readonly reason: CertifiedSafeStateReason;
    };

export interface CertifiedAllocationKernelInput {
  readonly generatedAt?: string;
  readonly source?: string;
  readonly plan: CertifiedCalibrationPlan;
  readonly points: readonly CertifiedCalibrationPointLedger[];
  readonly modelSolved: boolean;
  readonly frame: CertifiedObservationFrame;
  readonly candidateIds: readonly number[];
  readonly budget: CertifiedLoopBudget;
}

export interface CertifiedAllocationKernelAggregate {
  readonly requiredSamples: number;
  readonly acceptedSamples: number;
  readonly rejectedSamples: number;
  readonly observedSamples: number;
  readonly deficit: number;
  readonly conserved: boolean;
  readonly terminated: boolean;
  readonly ready: boolean;
  readonly nextPointId: string | null;
  readonly confidenceFloorSatisfied: boolean;
  readonly withinBudget: boolean;
}

export interface CertifiedAllocationKernelBundle {
  readonly generatedAt: string;
  readonly source: string;
  readonly theoremIds: readonly CertifiedAllocationKernelTheoremId[];
  readonly observationZone: ObservationZone;
  readonly inputZone: ObservationZone;
  readonly inputZoneConsistent: boolean;
  readonly decision: CertifiedAllocationDecision;
  readonly aggregate: CertifiedAllocationKernelAggregate;
}

const MAX_MILLI_COORD = 999;

function nowIso(): string {
  return new Date().toISOString();
}

function clampNonNegativeInt(value: number): number {
  try {
    if (!Number.isFinite(value)) {
      return 0;
    }
    return Math.max(0, Math.floor(value));

  } catch (error) {
    throw new Error(`[clampNonNegativeInt] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function clampMilliCoord(value: number): number {
  return Math.min(MAX_MILLI_COORD, clampNonNegativeInt(value));
}

function normalizePoint(
  point: NormalizedMilliPoint | null
): NormalizedMilliPoint | null {
  try {
    if (point === null) {
      return null;
    }
    return {
      x: clampMilliCoord(point.x),
      y: clampMilliCoord(point.y),
    };

  } catch (error) {
    throw new Error(`[normalizePoint] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function normalizePlan(plan: CertifiedCalibrationPlan): CertifiedCalibrationPlan {
  return {
    pointIds: plan.pointIds.map((pointId) => pointId),
    samplesPerPoint: clampNonNegativeInt(plan.samplesPerPoint),
    maxAttemptsPerPoint: clampNonNegativeInt(plan.maxAttemptsPerPoint),
  };
}

function normalizePointLedger(
  point: CertifiedCalibrationPointLedger
): CertifiedCalibrationPointLedger {
  return {
    pointId: point.pointId,
    accepted: clampNonNegativeInt(point.accepted),
    rejected: clampNonNegativeInt(point.rejected),
    observed: clampNonNegativeInt(point.observed),
  };
}

function normalizeBudget(budget: CertifiedLoopBudget): CertifiedLoopBudget {
  return {
    deadlineMs: clampNonNegativeInt(budget.deadlineMs),
    elapsedMs: clampNonNegativeInt(budget.elapsedMs),
  };
}

function normalizeFrame(
  frame: CertifiedObservationFrame
): CertifiedObservationFrame {
  const gazePoint = normalizePoint(frame.gazePoint);
  return {
    gazePoint,
    zone: classifyObservationZone(gazePoint),
    confidenceScore: clampNonNegativeInt(frame.confidenceScore),
    confidenceFloor: clampNonNegativeInt(frame.confidenceFloor),
  };
}

function orderedPointLedgers(
  plan: CertifiedCalibrationPlan,
  points: readonly CertifiedCalibrationPointLedger[]
): CertifiedCalibrationPointLedger[] {
  const pointMap = new Map(
    points.map((point) => {
      const normalized = normalizePointLedger(point);
      return [normalized.pointId, normalized] as const;
    })
  );

  return plan.pointIds.map(
    (pointId) =>
      pointMap.get(pointId) ?? {
        pointId,
        accepted: 0,
        rejected: 0,
        observed: 0,
      }
  );
}

/**
 * Handles the gnosis observation Column workflow.
 */
export function observationColumn(point: NormalizedMilliPoint): number {
  return Math.floor((clampMilliCoord(point.x) * 3) / 1000);
}

/**
 * Handles the gnosis observation Row workflow.
 */
export function observationRow(point: NormalizedMilliPoint): number {
  return Math.floor((clampMilliCoord(point.y) * 3) / 1000);
}

/**
 * Handles the gnosis classify Visible Zone workflow.
 */
export function classifyVisibleZone(
  point: NormalizedMilliPoint
): AdmittedObservationZone {
  try {
    const row = observationRow(point);
    const column = observationColumn(point);

    if (row === 0) {
      if (column === 0) {
        return 'top-left';
      }
      if (column === 1) {
        return 'top-center';
      }
      return 'top-right';
    }

    if (row === 1) {
      if (column === 0) {
        return 'center-left';
      }
      if (column === 1) {
        return 'center';
      }
      return 'center-right';
    }

    if (column === 0) {
      return 'bottom-left';
    }
    if (column === 1) {
      return 'bottom-center';
    }
    return 'bottom-right';

  } catch (error) {
    throw new Error(`[classifyVisibleZone] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Handles the gnosis classify Observation Zone workflow.
 */
export function classifyObservationZone(
  point: NormalizedMilliPoint | null
): ObservationZone {
  return point === null ? 'unavailable' : classifyVisibleZone(point);
}

/**
 * Handles the gnosis confidence Floor Satisfied workflow.
 */
export function confidenceFloorSatisfied(
  frame: CertifiedObservationFrame
): boolean {
  return (
    frame.zone === 'unavailable' ||
    frame.confidenceScore >= frame.confidenceFloor
  );
}

/**
 * Handles the gnosis point Deficit workflow.
 */
export function pointDeficit(
  plan: CertifiedCalibrationPlan,
  point: CertifiedCalibrationPointLedger
): number {
  return Math.max(0, plan.samplesPerPoint - point.accepted);
}

/**
 * Handles the gnosis point Observation Conserved workflow.
 */
export function pointObservationConserved(
  point: CertifiedCalibrationPointLedger
): boolean {
  return point.observed === point.accepted + point.rejected;
}

/**
 * Handles the gnosis accepted Sample Count workflow.
 */
export function acceptedSampleCount(
  points: readonly CertifiedCalibrationPointLedger[]
): number {
  return points.reduce((sum, point) => sum + point.accepted, 0);
}

/**
 * Handles the gnosis rejected Sample Count workflow.
 */
export function rejectedSampleCount(
  points: readonly CertifiedCalibrationPointLedger[]
): number {
  return points.reduce((sum, point) => sum + point.rejected, 0);
}

/**
 * Handles the gnosis observed Sample Count workflow.
 */
export function observedSampleCount(
  points: readonly CertifiedCalibrationPointLedger[]
): number {
  return points.reduce((sum, point) => sum + point.observed, 0);
}

/**
 * Handles the gnosis required Samples workflow.
 */
export function requiredSamples(plan: CertifiedCalibrationPlan): number {
  return plan.pointIds.length * plan.samplesPerPoint;
}

/**
 * Handles the gnosis calibration Deficit workflow.
 */
export function calibrationDeficit(
  plan: CertifiedCalibrationPlan,
  points: readonly CertifiedCalibrationPointLedger[]
): number {
  return Math.max(0, requiredSamples(plan) - acceptedSampleCount(points));
}

/**
 * Handles the gnosis calibration Observation Conserved workflow.
 */
export function calibrationObservationConserved(
  points: readonly CertifiedCalibrationPointLedger[]
): boolean {
  return points.every((point) => pointObservationConserved(point));
}

/**
 * Handles the gnosis point Complete workflow.
 */
export function pointComplete(
  plan: CertifiedCalibrationPlan,
  point: CertifiedCalibrationPointLedger
): boolean {
  return pointDeficit(plan, point) === 0;
}

/**
 * Handles the gnosis point Exhausted workflow.
 */
export function pointExhausted(
  plan: CertifiedCalibrationPlan,
  point: CertifiedCalibrationPointLedger
): boolean {
  return (
    point.observed >= plan.maxAttemptsPerPoint && pointDeficit(plan, point) > 0
  );
}

/**
 * Handles the gnosis point Terminal workflow.
 */
export function pointTerminal(
  plan: CertifiedCalibrationPlan,
  point: CertifiedCalibrationPointLedger
): boolean {
  return pointComplete(plan, point) || pointExhausted(plan, point);
}

/**
 * Handles the gnosis calibration Terminated workflow.
 */
export function calibrationTerminated(
  plan: CertifiedCalibrationPlan,
  points: readonly CertifiedCalibrationPointLedger[]
): boolean {
  return orderedPointLedgers(plan, points).every((point) => pointTerminal(plan, point));
}

/**
 * Handles the gnosis next Pending Point Id workflow.
 */
export function nextPendingPointId(
  plan: CertifiedCalibrationPlan,
  points: readonly CertifiedCalibrationPointLedger[]
): string | null {
  try {
    for (const point of orderedPointLedgers(plan, points)) {
      if (!pointTerminal(plan, point)) {
        return point.pointId;
      }
    }
    return null;

  } catch (error) {
    throw new Error(`[nextPendingPointId] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Handles the gnosis calibration Ready workflow.
 */
export function calibrationReady(
  plan: CertifiedCalibrationPlan,
  points: readonly CertifiedCalibrationPointLedger[],
  modelSolved: boolean
): boolean {
  return (
    calibrationTerminated(plan, points) &&
    calibrationDeficit(plan, points) === 0 &&
    modelSolved
  );
}

/**
 * Handles the gnosis within Budget workflow.
 */
export function withinBudget(budget: CertifiedLoopBudget): boolean {
  return budget.elapsedMs <= budget.deadlineMs;
}

/**
 * Handles the gnosis canonical Candidate Id workflow.
 */
export function canonicalCandidateId(
  candidateIds: readonly number[]
): number | null {
  try {
    if (candidateIds.length === 0) {
      return null;
    }

    let minimum = clampNonNegativeInt(candidateIds[0] ?? 0);
    for (let index = 1; index < candidateIds.length; index += 1) {
      const candidateId = clampNonNegativeInt(candidateIds[index] ?? 0);
      if (candidateId < minimum) {
        minimum = candidateId;
      }
    }
    return minimum;

  } catch (error) {
    throw new Error(`[canonicalCandidateId] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Handles the gnosis certified Allocation Step workflow.
 */
export function certifiedAllocationStep(
  plan: CertifiedCalibrationPlan,
  points: readonly CertifiedCalibrationPointLedger[],
  modelSolved: boolean,
  frame: CertifiedObservationFrame,
  candidateIds: readonly number[],
  budget: CertifiedLoopBudget
): CertifiedAllocationDecision {
  try {
    if (!withinBudget(budget)) {
      return {
        kind: 'safe-state',
        reason: 'deadline-exceeded',
      };
    }

    if (!calibrationReady(plan, points, modelSolved)) {
      return {
        kind: 'safe-state',
        reason: 'not-calibrated',
      };
    }

    if (frame.zone === 'unavailable') {
      return {
        kind: 'safe-state',
        reason: 'observation-unavailable',
      };
    }

    const candidateId = canonicalCandidateId(candidateIds);
    if (candidateId === null) {
      return {
        kind: 'safe-state',
        reason: 'no-candidates',
      };
    }

    return {
      kind: 'allocate',
      candidateId,
    };

  } catch (error) {
    throw new Error(`[certifiedAllocationStep] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Handles the gnosis derive Certified Allocation Kernel workflow.
 */
export function deriveCertifiedAllocationKernel(
  input: CertifiedAllocationKernelInput
): CertifiedAllocationKernelBundle {
  const plan = normalizePlan(input.plan);
  const points = orderedPointLedgers(plan, input.points);
  const frame = normalizeFrame(input.frame);
  const budget = normalizeBudget(input.budget);
  const decision = certifiedAllocationStep(
    plan,
    points,
    input.modelSolved,
    frame,
    input.candidateIds,
    budget
  );

  return {
    generatedAt: input.generatedAt ?? nowIso(),
    source: input.source?.trim() || 'certified-allocation-kernel',
    theoremIds: CERTIFIED_ALLOCATION_KERNEL_THEOREM_IDS,
    observationZone: frame.zone,
    inputZone: input.frame.zone,
    inputZoneConsistent: input.frame.zone === frame.zone,
    decision,
    aggregate: {
      requiredSamples: requiredSamples(plan),
      acceptedSamples: acceptedSampleCount(points),
      rejectedSamples: rejectedSampleCount(points),
      observedSamples: observedSampleCount(points),
      deficit: calibrationDeficit(plan, points),
      conserved: calibrationObservationConserved(points),
      terminated: calibrationTerminated(plan, points),
      ready: calibrationReady(plan, points, input.modelSolved),
      nextPointId: nextPendingPointId(plan, points),
      confidenceFloorSatisfied: confidenceFloorSatisfied(frame),
      withinBudget: withinBudget(budget),
    },
  };
}
