// @a0n/.ignored_gnosis/src/certified-app.ts
// and avoid top-level await (which breaks CJS output via esbuild).
// Functions that need them will fail at runtime in browser, but the module
// can be imported without error.

type RuntimeRequire = (specifier: string) => unknown;
type RuntimeCreateRequire = (baseUrl: string) => RuntimeRequire;
type NodeProcessWithBuiltinModule = NodeJS.Process & {
  getBuiltinModule?: (id: string) => unknown;
};
type NodeModuleBuiltin = {
  createRequire?: RuntimeCreateRequire;
};

function resolveRuntimeRequire(): RuntimeRequire | null {
  if (typeof require !== 'undefined') {
    return require as RuntimeRequire;
  }

  const runtimeProcess =
    typeof process !== 'undefined'
      ? (process as NodeProcessWithBuiltinModule)
      : null;
  const nodeModule = runtimeProcess?.getBuiltinModule?.(
    'module'
  ) as NodeModuleBuiltin | null | undefined;
  const createRequire =
    typeof nodeModule?.createRequire === 'function'
      ? nodeModule.createRequire
      : null;

  if (!createRequire || typeof import.meta === 'undefined' || !import.meta.url) {
    return null;
  }

  return createRequire(import.meta.url);
}

const nodeRequire = resolveRuntimeRequire();

function requireNodeModule<T>(specifier: string): T {
  if (!nodeRequire) {
    throw new Error(
      `Certified app filesystem integration requires a Node-compatible module loader for ${specifier}.`
    );
  }

  return nodeRequire(specifier) as T;
}

let _fsCache: typeof import('node:fs') | null = null;
// async function _getFs(): Promise<typeof import('node:fs')> {
//   if (!_fsCache) {
//     _fsCache = await import('node:fs');
//   }
//   return _fsCache;
// }
function getFsSync(): typeof import('node:fs') {
  try {
    if (!_fsCache) {
      _fsCache = requireNodeModule<typeof import('node:fs')>('node:fs');
    }
    return _fsCache;

  } catch (error) {
    throw new Error(`[getFsSync] ${error instanceof Error ? error.message : String(error)}`);
  }
}

let _pathCache: typeof import('node:path') | null = null;
function getPathSync(): typeof import('node:path') {
  try {
    if (!_pathCache) {
      _pathCache = requireNodeModule<typeof import('node:path')>('node:path');
    }
    return _pathCache;

  } catch (error) {
    throw new Error(`[getPathSync] ${error instanceof Error ? error.message : String(error)}`);
  }
}

let _parseGgProgram: typeof import('@a0n/aeon-logic')['parseGgProgram'] | null =
  null;
// async function _getParseGgProgram(): Promise<(typeof import('@a0n/aeon-logic'))['parseGgProgram']> {
//   if (!_parseGgProgram) {
//     const mod = await import('@a0n/aeon-logic');
//     _parseGgProgram = mod.parseGgProgram;
//   }
//   return _parseGgProgram;
// }
function getParseGgProgramSync(): typeof import('@a0n/aeon-logic')['parseGgProgram'] {
  try {
    if (!_parseGgProgram) {
      const mod = requireNodeModule<typeof import('@a0n/aeon-logic')>(
        '@a0n/aeon-logic'
      );
      _parseGgProgram = mod.parseGgProgram;
    }
    return _parseGgProgram;

  } catch (error) {
    throw new Error(`[getParseGgProgramSync] ${error instanceof Error ? error.message : String(error)}`);
  }
}
import type { RuntimeTarget } from './capabilities/index';
import {
  inferCapabilitiesFromGgSource,
  summarizeCapabilityRequirements,
} from './capabilities/index';
import { QDoc } from './crdt/qdoc';
import { certifyGnosisProgram, toLawCapabilitySummary } from './law';
import { lowerUfcsSource } from './ufcs';
import {
  auditAgentSwarmCertificate,
  type AgentSwarmMergeCertificate,
} from './runtime/agent-swarm-contract';
import {
  auditUniverseCertificate,
  sha256Hex,
  stableStringify,
  type LawCapabilitySummary,
  type UniverseCertificate,
} from './universe/index';

const CERTIFIED_APP_ENVELOPES_MAP = 'certified_app_envelopes';
const CERTIFIED_APP_ENVELOPE_ALIASES_MAP = 'certified_app_envelope_aliases';
const CERTIFIED_APP_CURRENT_ALIAS = 'current';

export type CertifiedAppAttestationKind =
  | 'vfs_snapshot'
  | 'deploy_build'
  | 'runtime_start'
  | 'publish_proof'
  | 'aeonpid_registration'
  | 'justice_gate'
  | 'agent_swarm_adjudication';

export interface CertifiedAppEnvelopeCore {
  readonly appId: string;
  readonly sourcePath: string;
  readonly entrypoint: string;
  readonly sourceDigest: string;
  readonly artifactId: string;
  readonly artifactFingerprint: string;
  readonly runtimeTarget: RuntimeTarget;
  readonly capabilitySummary: LawCapabilitySummary;
  readonly dependencyCertificateDigests: readonly string[];
  readonly universeDigest: string;
  readonly universeCertificate: UniverseCertificate;
}

interface CertifiedAppAttestationBase {
  readonly kind: CertifiedAppAttestationKind;
  readonly attestedAt: string;
  readonly actor?: string;
}

export interface CertifiedAppVfsSnapshotAttestation
  extends CertifiedAppAttestationBase {
  readonly kind: 'vfs_snapshot';
  readonly snapshotDigest: string;
  readonly sourceDigest: string;
  readonly sourceRoot: string;
  readonly sourceMode: 'vfs' | 'filesystem';
  readonly fileCount: number;
  readonly materializedRoot?: string;
}

export interface CertifiedAppDeployBuildAttestation
  extends CertifiedAppAttestationBase {
  readonly kind: 'deploy_build';
  readonly buildId: string;
  readonly commitHash: string;
  readonly commitShort: string;
  readonly branch: string;
  readonly version: string;
  readonly deployTimestamp: string;
  readonly deployEpoch: number;
}

export interface CertifiedAppRuntimeStartAttestation
  extends CertifiedAppAttestationBase {
  readonly kind: 'runtime_start';
  readonly runtimeName: string;
  readonly hostTarget: RuntimeTarget;
  readonly pid?: string;
  readonly port?: number;
  readonly healthUrl?: string;
}

export interface CertifiedAppPublishProofAttestation
  extends CertifiedAppAttestationBase {
  readonly kind: 'publish_proof';
  readonly manifestPath: string;
  readonly proofMode: 'anchor' | 'snapshot' | 'unknown';
  readonly app?: string;
  readonly site?: string;
  readonly domain?: string;
  readonly tokenId?: string | null;
  readonly aeonPid?: {
    readonly cid: string;
    readonly instance: string;
    readonly owner: string;
    readonly name: string;
    readonly generation: number;
  } | null;
}

export interface CertifiedAppAeonPidRegistrationAttestation
  extends CertifiedAppAttestationBase {
  readonly kind: 'aeonpid_registration';
  readonly directoryId: string;
  readonly app: string;
  readonly site: string;
  readonly domain: string;
  readonly tokenId?: string | null;
  readonly proofMode: 'anchor' | 'snapshot' | 'unknown';
}

export interface CertifiedAppJusticeGateAttestation
  extends CertifiedAppAttestationBase {
  readonly kind: 'justice_gate';
  readonly allowed: boolean;
  readonly activeHold: boolean;
  readonly overrideGranted: boolean;
  readonly reason?: string;
}

export interface CertifiedAppAgentSwarmAdjudicationSummary {
  readonly certificateDigest: string;
  readonly changeId: string;
  readonly targetBranch: string;
  readonly protectedBranch: boolean;
  readonly riskLevel: AgentSwarmMergeCertificate['riskLevel'];
  readonly decision: AgentSwarmMergeCertificate['decision'];
  readonly issueCount: number;
  readonly errorCount: number;
  readonly warningCount: number;
  readonly infoCount: number;
  readonly taskCount: number;
  readonly executionCount: number;
  readonly approvalCount: number;
  readonly humanApprovalCount: number;
}

export interface CertifiedAppAgentSwarmAdjudicationAttestation
  extends CertifiedAppAttestationBase {
  readonly kind: 'agent_swarm_adjudication';
  readonly certificate: AgentSwarmMergeCertificate;
  readonly summary: CertifiedAppAgentSwarmAdjudicationSummary;
}

export type CertifiedAppEnvelopeAttestation =
  | CertifiedAppVfsSnapshotAttestation
  | CertifiedAppDeployBuildAttestation
  | CertifiedAppRuntimeStartAttestation
  | CertifiedAppPublishProofAttestation
  | CertifiedAppAeonPidRegistrationAttestation
  | CertifiedAppJusticeGateAttestation
  | CertifiedAppAgentSwarmAdjudicationAttestation;

export interface CertifiedAppEnvelope {
  readonly version: 1;
  readonly envelopeId: string;
  readonly digest: string;
  readonly createdAt: string;
  readonly updatedAt: string;
  readonly core: CertifiedAppEnvelopeCore;
  readonly attestations: readonly CertifiedAppEnvelopeAttestation[];
}

export interface CertifiedAppEnvelopeMirror {
  readonly version: 1;
  readonly envelopeId: string;
  readonly digest: string;
  readonly createdAt: string;
  readonly updatedAt: string;
  readonly core: CertifiedAppEnvelopeCore;
  readonly attestationKinds: readonly CertifiedAppAttestationKind[];
  readonly attestations: readonly CertifiedAppEnvelopeAttestation[];
}

export interface CertifiedAppEnvelopeAudit {
  readonly selfConsistent: boolean;
  readonly issues: readonly string[];
}

export interface CreateCertifiedAppEnvelopeInput {
  readonly appId: string;
  readonly sourcePath: string;
  readonly entrypoint: string;
  readonly sourceDigest?: string;
  readonly artifactId: string;
  readonly artifactFingerprint: string;
  readonly runtimeTarget: RuntimeTarget;
  readonly universeCertificate: UniverseCertificate;
  readonly capabilitySummary?: LawCapabilitySummary;
  readonly dependencyCertificateDigests?: readonly string[];
  readonly createdAt?: string;
  readonly attestations?: readonly CertifiedAppEnvelopeAttestation[];
}

export interface CreateCertifiedAppEnvelopeFromGgSourceInput
  extends Omit<
    CreateCertifiedAppEnvelopeInput,
    'universeCertificate' | 'capabilitySummary'
  > {
  readonly ggSource: string;
}

function assertNonEmpty(value: string, field: string): string {
  try {
    const normalized = value.trim();
    if (normalized.length === 0) {
      throw new Error(`"${field}" cannot be empty`);
    }
    return normalized;

  } catch (error) {
    throw new Error(`[assertNonEmpty] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function normalizeDigestList(values: readonly string[] | undefined): string[] {
  return [...(values ?? [])]
    .map((value) => value.trim())
    .filter((value) => value.length > 0)
    .sort();
}

function toIsoTimestamp(value: string | undefined, field: string): string {
  try {
    const candidate = value ?? new Date().toISOString();
    if (Number.isNaN(Date.parse(candidate))) {
      throw new Error(`"${field}" must be a valid ISO timestamp`);
    }
    return candidate;

  } catch (error) {
    throw new Error(`[toIsoTimestamp] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function buildCapabilitySummary(
  runtimeTarget: RuntimeTarget,
  universeCertificate: UniverseCertificate,
  capabilitySummary?: LawCapabilitySummary
): LawCapabilitySummary {
  const summary = capabilitySummary ?? universeCertificate.capabilitySummary;
  return {
    ...(summary.target ? { target: summary.target } : {}),
    ...(summary.required ? { required: [...summary.required] } : {}),
    ...(summary.declared ? { declared: [...summary.declared] } : {}),
    ...(summary.inferred ? { inferred: [...summary.inferred] } : {}),
    ...(summary.importedSurface
      ? { importedSurface: [...summary.importedSurface] }
      : {}),
    ...(summary.issueCount !== undefined
      ? { issueCount: summary.issueCount }
      : {}),
    target: summary.target ?? runtimeTarget,
  };
}

function canonicalizeAttestation(
  attestation: CertifiedAppEnvelopeAttestation
): CertifiedAppEnvelopeAttestation {
  try {
    switch (attestation.kind) {
      case 'vfs_snapshot':
        return {
          kind: 'vfs_snapshot',
          attestedAt: toIsoTimestamp(attestation.attestedAt, 'attestedAt'),
          sourceMode: attestation.sourceMode,
          snapshotDigest: assertNonEmpty(
            attestation.snapshotDigest,
            'snapshotDigest'
          ),
          sourceDigest: assertNonEmpty(attestation.sourceDigest, 'sourceDigest'),
          sourceRoot: assertNonEmpty(attestation.sourceRoot, 'sourceRoot'),
          fileCount: attestation.fileCount,
          ...(attestation.materializedRoot
            ? {
                materializedRoot: assertNonEmpty(
                  attestation.materializedRoot,
                  'materializedRoot'
                ),
              }
            : {}),
          ...(attestation.actor ? { actor: attestation.actor.trim() } : {}),
        };
      case 'deploy_build':
        return {
          kind: 'deploy_build',
          attestedAt: toIsoTimestamp(attestation.attestedAt, 'attestedAt'),
          buildId: assertNonEmpty(attestation.buildId, 'buildId'),
          commitHash: attestation.commitHash.trim(),
          commitShort: attestation.commitShort.trim(),
          branch: attestation.branch.trim(),
          version: attestation.version.trim(),
          deployTimestamp: toIsoTimestamp(
            attestation.deployTimestamp,
            'deployTimestamp'
          ),
          deployEpoch: attestation.deployEpoch,
          ...(attestation.actor ? { actor: attestation.actor.trim() } : {}),
        };
      case 'runtime_start':
        return {
          kind: 'runtime_start',
          attestedAt: toIsoTimestamp(attestation.attestedAt, 'attestedAt'),
          runtimeName: assertNonEmpty(attestation.runtimeName, 'runtimeName'),
          hostTarget: attestation.hostTarget,
          ...(attestation.pid ? { pid: attestation.pid.trim() } : {}),
          ...(attestation.port !== undefined ? { port: attestation.port } : {}),
          ...(attestation.healthUrl
            ? { healthUrl: attestation.healthUrl.trim() }
            : {}),
          ...(attestation.actor ? { actor: attestation.actor.trim() } : {}),
        };
      case 'publish_proof':
        return {
          kind: 'publish_proof',
          attestedAt: toIsoTimestamp(attestation.attestedAt, 'attestedAt'),
          manifestPath: assertNonEmpty(attestation.manifestPath, 'manifestPath'),
          proofMode: attestation.proofMode,
          ...(attestation.app ? { app: attestation.app.trim() } : {}),
          ...(attestation.site ? { site: attestation.site.trim() } : {}),
          ...(attestation.domain ? { domain: attestation.domain.trim() } : {}),
          ...(attestation.tokenId !== undefined
            ? { tokenId: attestation.tokenId }
            : {}),
          ...(attestation.aeonPid !== undefined
            ? { aeonPid: attestation.aeonPid }
            : {}),
          ...(attestation.actor ? { actor: attestation.actor.trim() } : {}),
        };
      case 'aeonpid_registration':
        return {
          kind: 'aeonpid_registration',
          attestedAt: toIsoTimestamp(attestation.attestedAt, 'attestedAt'),
          directoryId: assertNonEmpty(attestation.directoryId, 'directoryId'),
          app: assertNonEmpty(attestation.app, 'app'),
          site: assertNonEmpty(attestation.site, 'site'),
          domain: assertNonEmpty(attestation.domain, 'domain'),
          proofMode: attestation.proofMode,
          ...(attestation.tokenId !== undefined
            ? { tokenId: attestation.tokenId }
            : {}),
          ...(attestation.actor ? { actor: attestation.actor.trim() } : {}),
        };
      case 'justice_gate':
        return {
          kind: 'justice_gate',
          attestedAt: toIsoTimestamp(attestation.attestedAt, 'attestedAt'),
          allowed: attestation.allowed,
          activeHold: attestation.activeHold,
          overrideGranted: attestation.overrideGranted,
          ...(attestation.reason ? { reason: attestation.reason.trim() } : {}),
          ...(attestation.actor ? { actor: attestation.actor.trim() } : {}),
        };
      case 'agent_swarm_adjudication': {
        const expectedSummary = buildAgentSwarmAdjudicationSummary(
          attestation.certificate
        );
        const attestedAt = toIsoTimestamp(attestation.attestedAt, 'attestedAt');
        if (Date.parse(attestedAt) < attestation.certificate.generatedAt) {
          throw new Error(
            'agent_swarm_adjudication attestedAt cannot be earlier than the certificate generatedAt timestamp'
          );
        }
        if (
          stableStringify(attestation.summary) !==
          stableStringify(expectedSummary)
        ) {
          throw new Error(
            'agent_swarm_adjudication summary does not match the canonical certificate summary'
          );
        }
        return {
          kind: 'agent_swarm_adjudication',
          attestedAt,
          certificate: attestation.certificate,
          summary: expectedSummary,
          ...(attestation.actor ? { actor: attestation.actor.trim() } : {}),
        };
      }
    }

  } catch (error) {
    throw new Error(`[canonicalizeAttestation] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function countAgentSwarmIssuesBySeverity(
  issues: AgentSwarmMergeCertificate['issues']
): {
  readonly infoCount: number;
  readonly warningCount: number;
  readonly errorCount: number;
} {
  try {
    let infoCount = 0;
    let warningCount = 0;
    let errorCount = 0;

    for (const issue of issues) {
      if (issue.severity === 'info') {
        infoCount += 1;
      } else if (issue.severity === 'warning') {
        warningCount += 1;
      } else {
        errorCount += 1;
      }
    }

    return {
      infoCount,
      warningCount,
      errorCount,
    };

  } catch (error) {
    throw new Error(`[countAgentSwarmIssuesBySeverity] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function buildAgentSwarmAdjudicationSummary(
  certificate: AgentSwarmMergeCertificate
): CertifiedAppAgentSwarmAdjudicationSummary {
  const { infoCount, warningCount, errorCount } =
    countAgentSwarmIssuesBySeverity(certificate.issues);
  return {
    certificateDigest: certificate.digest,
    changeId: certificate.changeId,
    targetBranch: certificate.targetBranch,
    protectedBranch: certificate.protectedBranch,
    riskLevel: certificate.riskLevel,
    decision: certificate.decision,
    issueCount: certificate.issues.length,
    errorCount,
    warningCount,
    infoCount,
    taskCount: certificate.metrics.taskCount,
    executionCount: certificate.metrics.executionCount,
    approvalCount: certificate.metrics.approvalCount,
    humanApprovalCount: certificate.metrics.humanApprovalCount,
  };
}

function buildEnvelopeId(core: CertifiedAppEnvelopeCore): string {
  return sha256Hex(stableStringify(normalizeEnvelopeCoreForDigest(core)));
}

function buildEnvelopeDigest(
  envelope: Omit<CertifiedAppEnvelope, 'digest'>
): string {
  return sha256Hex(stableStringify(normalizeEnvelopeForDigest(envelope)));
}

function normalizeUniverseCertificateForDigest(
  certificate: UniverseCertificate
): Record<string, unknown> {
  const normalized = {
    ...certificate,
    buildMetadata: {
      ...certificate.buildMetadata,
      verifiedAt: 0,
    },
    clockwork: {
      ...certificate.clockwork,
      verifiedAt: 0,
    },
    ...(certificate.runtimeMetadata
      ? {
          runtimeMetadata: {
            ...certificate.runtimeMetadata,
            executedAt: 0,
          },
        }
      : {}),
  };
  const { digest: _ignoredDigest, ...withoutDigest } = normalized;
  return {
    ...withoutDigest,
    digest: sha256Hex(stableStringify(withoutDigest)),
  };
}

function normalizeEnvelopeCoreForDigest(
  core: CertifiedAppEnvelopeCore
): Record<string, unknown> {
  const normalizedCertificate = normalizeUniverseCertificateForDigest(
    core.universeCertificate
  );
  return {
    ...core,
    universeDigest: (normalizedCertificate as { readonly digest: string })
      .digest,
    universeCertificate: normalizedCertificate,
  };
}

function normalizeEnvelopeForDigest(
  envelope: Omit<CertifiedAppEnvelope, 'digest'>
): Record<string, unknown> {
  return {
    ...envelope,
    core: normalizeEnvelopeCoreForDigest(envelope.core),
  };
}

function createEnvelopeCore(
  input: CreateCertifiedAppEnvelopeInput
): CertifiedAppEnvelopeCore {
  const dependencyCertificateDigests = normalizeDigestList(
    input.dependencyCertificateDigests ??
      input.universeCertificate.dependencyCertificateDigests
  );
  const capabilitySummary = buildCapabilitySummary(
    input.runtimeTarget,
    input.universeCertificate,
    input.capabilitySummary
  );

  return {
    appId: assertNonEmpty(input.appId, 'appId'),
    sourcePath: assertNonEmpty(input.sourcePath, 'sourcePath'),
    entrypoint: assertNonEmpty(input.entrypoint, 'entrypoint'),
    sourceDigest: assertNonEmpty(
      input.sourceDigest ?? input.universeCertificate.sourceDigest,
      'sourceDigest'
    ),
    artifactId: assertNonEmpty(input.artifactId, 'artifactId'),
    artifactFingerprint: assertNonEmpty(
      input.artifactFingerprint,
      'artifactFingerprint'
    ),
    runtimeTarget: input.runtimeTarget,
    capabilitySummary,
    dependencyCertificateDigests,
    universeDigest: input.universeCertificate.digest,
    universeCertificate: input.universeCertificate,
  };
}

/**
 * Creates the Certified App Envelope.
 */
export function createCertifiedAppEnvelope(
  input: CreateCertifiedAppEnvelopeInput
): CertifiedAppEnvelope {
  const createdAt = toIsoTimestamp(input.createdAt, 'createdAt');
  const core = createEnvelopeCore(input);
  const envelopeId = buildEnvelopeId(core);
  const attestations = [...(input.attestations ?? [])].map(
    canonicalizeAttestation
  );
  const latestAttestation = attestations[attestations.length - 1];
  const updatedAt = latestAttestation ? latestAttestation.attestedAt : createdAt;
  const withoutDigest = {
    version: 1 as const,
    envelopeId,
    createdAt,
    updatedAt,
    core,
    attestations,
  };

  return {
    ...withoutDigest,
    digest: buildEnvelopeDigest(withoutDigest),
  };
}

/**
 * Creates the Certified App Envelope From Gg Source.
 */
export function createCertifiedAppEnvelopeFromGgSource(
  input: CreateCertifiedAppEnvelopeFromGgSourceInput
): CertifiedAppEnvelope {
  const ggSource = assertNonEmpty(input.ggSource, 'ggSource');
  const normalizedSource = lowerUfcsSource(ggSource);
  const capabilitySummary = toLawCapabilitySummary(
    summarizeCapabilityRequirements(
      inferCapabilitiesFromGgSource(normalizedSource)
    ),
    input.runtimeTarget
  );
  const universeCertificate = certifyGnosisProgram(
    getParseGgProgramSync()(normalizedSource),
    {
      subject: input.sourcePath,
      ...(input.sourceDigest
        ? { sourceDigest: input.sourceDigest }
        : { source: normalizedSource }),
      capabilitySummary,
    }
  ).certificate;

  return createCertifiedAppEnvelope({
    ...input,
    sourceDigest: input.sourceDigest ?? universeCertificate.sourceDigest,
    capabilitySummary,
    universeCertificate,
  });
}

/**
 * Handles the gnosis append Certified App Envelope Attestation workflow.
 */
export function appendCertifiedAppEnvelopeAttestation(
  envelope: CertifiedAppEnvelope,
  attestation: CertifiedAppEnvelopeAttestation
): CertifiedAppEnvelope {
  const nextAttestation = canonicalizeAttestation(attestation);
  const attestations = [...envelope.attestations, nextAttestation];
  const withoutDigest = {
    version: envelope.version,
    envelopeId: envelope.envelopeId,
    createdAt: envelope.createdAt,
    updatedAt: nextAttestation.attestedAt,
    core: envelope.core,
    attestations,
  };

  return {
    ...withoutDigest,
    digest: buildEnvelopeDigest(withoutDigest),
  };
}

/**
 * Creates the Certified App Envelope Mirror.
 */
export function createCertifiedAppEnvelopeMirror(
  envelope: CertifiedAppEnvelope
): CertifiedAppEnvelopeMirror {
  return {
    version: envelope.version,
    envelopeId: envelope.envelopeId,
    digest: envelope.digest,
    createdAt: envelope.createdAt,
    updatedAt: envelope.updatedAt,
    core: envelope.core,
    attestationKinds: envelope.attestations.map(
      (attestation) => attestation.kind
    ),
    attestations: envelope.attestations,
  };
}

/**
 * Handles the gnosis serialize Certified App Envelope Mirror workflow.
 */
export function serializeCertifiedAppEnvelopeMirror(
  envelope: CertifiedAppEnvelope
): string {
  return `${JSON.stringify(
    createCertifiedAppEnvelopeMirror(envelope),
    null,
    2
  )}\n`;
}

function sortAttestationKinds(
  envelope: CertifiedAppEnvelope
): CertifiedAppEnvelope {
  return {
    ...envelope,
    attestations: [...envelope.attestations],
  };
}

/**
 * Handles the gnosis audit Certified App Envelope workflow.
 */
export function auditCertifiedAppEnvelope(
  envelope: CertifiedAppEnvelope
): CertifiedAppEnvelopeAudit {
  try {
    const issues: string[] = [];
    const universeAudit = auditUniverseCertificate(
      envelope.core.universeCertificate
    );
    if (!universeAudit.selfConsistent) {
      issues.push(...universeAudit.issues);
    }

    if (envelope.version !== 1) {
      issues.push(`Unknown certified app envelope version: ${envelope.version}`);
    }

    const expectedEnvelopeId = buildEnvelopeId(envelope.core);
    if (expectedEnvelopeId !== envelope.envelopeId) {
      issues.push('Certified app envelope id does not match the canonical core.');
    }

    const { digest: _ignoredDigest, ...withoutDigest } = envelope;
    const expectedDigest = buildEnvelopeDigest(withoutDigest);
    if (expectedDigest !== envelope.digest) {
      issues.push(
        'Certified app envelope digest does not match the envelope body.'
      );
    }

    if (
      envelope.core.universeDigest !== envelope.core.universeCertificate.digest
    ) {
      issues.push(
        'Universe digest does not match the embedded universe certificate.'
      );
    }

    if (
      envelope.core.sourceDigest !==
      envelope.core.universeCertificate.sourceDigest
    ) {
      issues.push(
        'Source digest does not match the embedded universe certificate.'
      );
    }

    const expectedDependencyDigests = normalizeDigestList(
      envelope.core.universeCertificate.dependencyCertificateDigests
    );
    const actualDependencyDigests = normalizeDigestList(
      envelope.core.dependencyCertificateDigests
    );
    if (
      stableStringify(expectedDependencyDigests) !==
      stableStringify(actualDependencyDigests)
    ) {
      issues.push(
        'Dependency certificate digests do not match the embedded universe certificate.'
      );
    }

    const capabilityTarget = envelope.core.capabilitySummary.target;
    if (
      capabilityTarget &&
      capabilityTarget.length > 0 &&
      capabilityTarget !== envelope.core.runtimeTarget
    ) {
      issues.push(
        `Capability target "${capabilityTarget}" does not match runtime target "${envelope.core.runtimeTarget}".`
      );
    }

    let previousAttestedAt = envelope.createdAt;
    for (const attestation of envelope.attestations) {
      if (Date.parse(attestation.attestedAt) < Date.parse(previousAttestedAt)) {
        issues.push(
          `Attestation "${attestation.kind}" is earlier than the previous envelope timestamp.`
        );
      }
      previousAttestedAt = attestation.attestedAt;

      if (
        attestation.kind === 'vfs_snapshot' &&
        attestation.sourceDigest !== envelope.core.sourceDigest
      ) {
        issues.push(
          'VFS snapshot attestation source digest does not match the certified source digest.'
        );
      }

      if (
        attestation.kind === 'runtime_start' &&
        attestation.hostTarget !== envelope.core.runtimeTarget
      ) {
        issues.push(
          `Runtime start host target "${attestation.hostTarget}" does not match certified runtime target "${envelope.core.runtimeTarget}".`
        );
      }

      if (
        attestation.kind === 'publish_proof' &&
        attestation.app &&
        attestation.app !== envelope.core.appId
      ) {
        issues.push(
          `Publish proof app "${attestation.app}" does not match certified app "${envelope.core.appId}".`
        );
      }

      if (
        attestation.kind === 'aeonpid_registration' &&
        attestation.app !== envelope.core.appId
      ) {
        issues.push(
          `AeonPID registration app "${attestation.app}" does not match certified app "${envelope.core.appId}".`
        );
      }
    }

    const publishProof = [...envelope.attestations]
      .reverse()
      .find(
        (attestation): attestation is CertifiedAppPublishProofAttestation =>
          attestation.kind === 'publish_proof'
      );
    const aeonPidRegistration = [...envelope.attestations]
      .reverse()
      .find(
        (
          attestation
        ): attestation is CertifiedAppAeonPidRegistrationAttestation =>
          attestation.kind === 'aeonpid_registration'
      );

    if (publishProof && aeonPidRegistration) {
      if (publishProof.site && publishProof.site !== aeonPidRegistration.site) {
        issues.push(
          `AeonPID registration site "${aeonPidRegistration.site}" does not match publish proof site "${publishProof.site}".`
        );
      }

      if (
        publishProof.domain &&
        publishProof.domain !== aeonPidRegistration.domain
      ) {
        issues.push(
          `AeonPID registration domain "${aeonPidRegistration.domain}" does not match publish proof domain "${publishProof.domain}".`
        );
      }

      if (
        publishProof.tokenId !== undefined &&
        publishProof.tokenId !== aeonPidRegistration.tokenId
      ) {
        issues.push(
          'AeonPID registration token id does not match the publish proof token id.'
        );
      }

      if (publishProof.proofMode !== aeonPidRegistration.proofMode) {
        issues.push(
          `AeonPID registration proof mode "${aeonPidRegistration.proofMode}" does not match publish proof mode "${publishProof.proofMode}".`
        );
      }
    }

    for (const attestation of envelope.attestations) {
      if (attestation.kind !== 'agent_swarm_adjudication') {
        continue;
      }

      const certificateAudit = auditAgentSwarmCertificate(
        attestation.certificate
      );
      if (!certificateAudit.selfConsistent) {
        issues.push(...certificateAudit.issues);
      }

      const expectedSummary = buildAgentSwarmAdjudicationSummary(
        attestation.certificate
      );
      if (
        stableStringify(attestation.summary) !== stableStringify(expectedSummary)
      ) {
        issues.push(
          'Agent swarm adjudication summary does not match the embedded merge certificate.'
        );
      }

      if (
        Date.parse(attestation.attestedAt) < attestation.certificate.generatedAt
      ) {
        issues.push(
          'Agent swarm adjudication attestedAt cannot be earlier than the embedded merge certificate timestamp.'
        );
      }
    }

    return {
      selfConsistent: issues.length === 0,
      issues,
    };

  } catch (error) {
    throw new Error(`[auditCertifiedAppEnvelope] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Handles the gnosis store Certified App Envelope In QDoc workflow.
 */
export function storeCertifiedAppEnvelopeInQDoc(
  doc: QDoc,
  envelope: CertifiedAppEnvelope,
  aliases: readonly string[] = []
): CertifiedAppEnvelope {
  try {
    const audit = auditCertifiedAppEnvelope(envelope);
    if (!audit.selfConsistent) {
      throw new Error(audit.issues.join(' '));
    }

    const envelopes = doc.getMap<CertifiedAppEnvelope>(
      CERTIFIED_APP_ENVELOPES_MAP
    );
    const aliasMap = doc.getMap<string>(CERTIFIED_APP_ENVELOPE_ALIASES_MAP);
    const canonicalEnvelope = sortAttestationKinds(envelope);

    envelopes.set(canonicalEnvelope.envelopeId, canonicalEnvelope);
    aliasMap.set(CERTIFIED_APP_CURRENT_ALIAS, canonicalEnvelope.envelopeId);

    for (const alias of [...aliases]
      .map((value) => value.trim())
      .filter((value) => value.length > 0)
      .sort()) {
      aliasMap.set(alias, canonicalEnvelope.envelopeId);
    }

    return canonicalEnvelope;

  } catch (error) {
    throw new Error(`[storeCertifiedAppEnvelopeInQDoc] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Loads the Certified App Envelope From QDoc.
 */
export function loadCertifiedAppEnvelopeFromQDoc(
  doc: QDoc,
  envelopeId: string
): CertifiedAppEnvelope | null {
  try {
    const normalizedEnvelopeId = envelopeId.trim();
    if (normalizedEnvelopeId.length === 0) {
      return null;
    }

    return (
      doc
        .getMap<CertifiedAppEnvelope>(CERTIFIED_APP_ENVELOPES_MAP)
        .get(normalizedEnvelopeId) ?? null
    );

  } catch (error) {
    throw new Error(`[loadCertifiedAppEnvelopeFromQDoc] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Loads the Certified App Envelope By Alias From QDoc.
 */
export function loadCertifiedAppEnvelopeByAliasFromQDoc(
  doc: QDoc,
  alias = CERTIFIED_APP_CURRENT_ALIAS
): CertifiedAppEnvelope | null {
  try {
    const normalizedAlias = alias.trim();
    if (normalizedAlias.length === 0) {
      return null;
    }

    const envelopeId =
      doc
        .getMap<string>(CERTIFIED_APP_ENVELOPE_ALIASES_MAP)
        .get(normalizedAlias) ?? normalizedAlias;
    return loadCertifiedAppEnvelopeFromQDoc(doc, envelopeId);

  } catch (error) {
    throw new Error(`[loadCertifiedAppEnvelopeByAliasFromQDoc] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Handles the gnosis append Certified App Envelope Attestation To QDoc workflow.
 */
export function appendCertifiedAppEnvelopeAttestationToQDoc(
  doc: QDoc,
  attestation: CertifiedAppEnvelopeAttestation,
  lookup = CERTIFIED_APP_CURRENT_ALIAS
): CertifiedAppEnvelope {
  try {
    const existing = loadCertifiedAppEnvelopeByAliasFromQDoc(doc, lookup);
    if (!existing) {
      throw new Error(`No certified app envelope found for "${lookup}".`);
    }

    const next = appendCertifiedAppEnvelopeAttestation(existing, attestation);
    return storeCertifiedAppEnvelopeInQDoc(doc, next, [lookup]);

  } catch (error) {
    throw new Error(`[appendCertifiedAppEnvelopeAttestationToQDoc] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Handles the gnosis save Certified App Envelope To File workflow.
 */
export function saveCertifiedAppEnvelopeToFile(
  targetPath: string,
  envelope: CertifiedAppEnvelope,
  aliases: readonly string[] = []
): void {
  const doc = new QDoc({ guid: `certified-app:${targetPath}` });
  storeCertifiedAppEnvelopeInQDoc(doc, envelope, aliases);
  const fs = getFsSync();
  const path = getPathSync();
  fs.mkdirSync(path.dirname(targetPath), { recursive: true });
  const tmpPath = `${targetPath}.tmp.${Date.now()}`;
  fs.writeFileSync(tmpPath, doc.encodeStateAsUpdate());
  fs.renameSync(tmpPath, targetPath);
}

/**
 * Loads the Certified App Envelope From File.
 */
export function loadCertifiedAppEnvelopeFromFile(
  targetPath: string,
  lookup = CERTIFIED_APP_CURRENT_ALIAS
): CertifiedAppEnvelope | null {
  try {
    const fs = getFsSync();
    if (!fs.existsSync(targetPath)) {
      return null;
    }

    const doc = new QDoc({ guid: `certified-app:${targetPath}` });
    doc.applyUpdate(new Uint8Array(fs.readFileSync(targetPath)), 'file');
    return loadCertifiedAppEnvelopeByAliasFromQDoc(doc, lookup);

  } catch (error) {
    throw new Error(`[loadCertifiedAppEnvelopeFromFile] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Handles the gnosis append Certified App Envelope Attestation To File workflow.
 */
export function appendCertifiedAppEnvelopeAttestationToFile(
  targetPath: string,
  attestation: CertifiedAppEnvelopeAttestation,
  lookup = CERTIFIED_APP_CURRENT_ALIAS
): CertifiedAppEnvelope {
  try {
    const existing = loadCertifiedAppEnvelopeFromFile(targetPath, lookup);
    if (!existing) {
      throw new Error(`No certified app envelope found for "${lookup}".`);
    }

    const next = appendCertifiedAppEnvelopeAttestation(existing, attestation);
    saveCertifiedAppEnvelopeToFile(targetPath, next, [lookup]);
    return next;

  } catch (error) {
    throw new Error(`[appendCertifiedAppEnvelopeAttestationToFile] ${error instanceof Error ? error.message : String(error)}`);
  }
}
