// @a0n/.ignored_gnosis/src/chaitin-omega.ts
/**
 * chaitin-omega.ts -- Chaitin's Omega as runtime void boundary
 *
 * Connects the formal theory (ChaitinOmega.lean, SolomonoffBuleyean.lean,
 * BuleyeanProbability.lean) to the runtime void walking machinery.
 *
 * ProgramSpace enumerates binary strings and classifies halting/non-halting.
 * Omega approximation computes the finite-prefix halting probability.
 * SolomonoffSpace maps Kolmogorov complexity to VoidBoundary rejection counts,
 * so complementDistribution() recovers the Solomonoff prior ordering.
 *
 * Zero modifications to void.ts -- everything composes through the standard
 * VoidBoundary interface.
 */

import {
  type VoidBoundary,
  createVoidBoundary,
  updateVoidBoundary,
  complementDistribution,
  createWalker,
  c0Choose,
  c0Update,
  c1Measure,
  c2c3Adapt,
} from './void';

// ============================================================================
// Types -- mirrors Lean structures
// ============================================================================

/** Classifier: given a binary string, returns true if it "halts". */
export type HaltingClassifier = (program: string) => boolean;

/**
 * ProgramSpace -- mirrors ChaitinOmega.lean
 * Enumeration of all binary strings up to maxLength.
 */
export interface ProgramSpace {
  alphabetSize: number;
  maxLength: number;
  totalPrograms: number;
  haltingPrograms: number;
  nonHaltingPrograms: number;
  /** The binary strings, indexed */
  programs: string[];
  /** Per-program halting classification */
  halting: boolean[];
}

/**
 * ComplexityAssignment -- mirrors SolomonoffBuleyean.lean
 * Per-choice Kolmogorov complexity upper bounds.
 */
export interface ComplexityAssignment {
  /** complexity[i] = K(choice_i), an upper bound on Kolmogorov complexity */
  complexity: number[];
  /** Ceiling: max complexity value used for weight computation */
  ceiling: number;
}

/**
 * SolomonoffSpace -- complexity assignment + empirical rounds
 */
export interface SolomonoffSpace {
  assignment: ComplexityAssignment;
  /** Number of empirical observation rounds (from Buleyean update) */
  rounds: number;
  /** Labels for each choice (optional) */
  labels?: string[];
}

/**
 * OmegaApproximation -- finite prefix Omega value
 */
export interface OmegaApproximation {
  numerator: number;
  denominator: number;
  ratio: number;
  nonHalting: number;
  prefixLength: number;
}

/**
 * OmegaConvergenceTrace -- sequence of approximations with monotonicity check
 */
export interface OmegaConvergenceTrace {
  approximations: OmegaApproximation[];
  monotoneIncreasing: boolean;
  finalRatio: number;
}

// ============================================================================
// Program space enumeration
// ============================================================================

/**
 * Enumerate all binary strings of lengths 1..maxLength.
 * Total programs = sum of alphabetSize^l for l = 1..maxLength.
 * Default maxLength = 12 (~4094 programs for binary alphabet).
 */
export function createProgramSpace(
  alphabetSize: number,
  maxLength: number,
  classifier: HaltingClassifier
): ProgramSpace {
  const programs: string[] = [];
  const halting: boolean[] = [];

  for (let len = 1; len <= maxLength; len++) {
    const count = alphabetSize ** len;
    for (let i = 0; i < count; i++) {
      const program = i.toString(alphabetSize).padStart(len, '0');
      programs.push(program);
      halting.push(classifier(program));
    }
  }

  const haltingCount = halting.filter(Boolean).length;
  return {
    alphabetSize,
    maxLength,
    totalPrograms: programs.length,
    haltingPrograms: haltingCount,
    nonHaltingPrograms: programs.length - haltingCount,
    programs,
    halting,
  };
}

// ============================================================================
// Omega approximation
// ============================================================================

/**
 * Compute the finite-prefix Omega approximation.
 * Omega_L = sum over halting programs p of alphabetSize^(-|p|)
 */
export function approximateOmega(ps: ProgramSpace): OmegaApproximation {
  try {
    let numerator = 0;
    let denominator = 0;

    for (let i = 0; i < ps.programs.length; i++) {
      const weight = ps.alphabetSize ** -ps.programs[i].length;
      denominator += weight;
      if (ps.halting[i]) {
        numerator += weight;
      }
    }

    return {
      numerator,
      denominator,
      ratio: denominator > 0 ? numerator / denominator : 0,
      nonHalting: ps.nonHaltingPrograms,
      prefixLength: ps.maxLength,
    };
  } catch (error) {
    throw new Error(
      `[approximateOmega] ${
        error instanceof Error ? error.message : String(error)
      }`
    );
  }
}

/**
 * Track convergence of Omega across increasing prefix lengths.
 * Verifies monotone increasing property (Omega_L <= Omega_{L+1}).
 */
export function trackOmegaConvergence(
  approximations: OmegaApproximation[]
): OmegaConvergenceTrace {
  try {
    let monotone = true;
    for (let i = 1; i < approximations.length; i++) {
      if (approximations[i].numerator < approximations[i - 1].numerator) {
        monotone = false;
        break;
      }
    }

    return {
      approximations,
      monotoneIncreasing: monotone,
      finalRatio:
        approximations.length > 0
          ? approximations[approximations.length - 1].ratio
          : 0,
    };
  } catch (error) {
    throw new Error(
      `[trackOmegaConvergence] ${
        error instanceof Error ? error.message : String(error)
      }`
    );
  }
}

// ============================================================================
// VoidBoundary bridges
// ============================================================================

/**
 * Convert ProgramSpace to VoidBoundary.
 * Non-halting programs become rejection counts: each non-halting program
 * at length l contributes alphabetSize^(-l) to the rejection count in
 * its dimension (program index).
 *
 * This creates a boundary over all programs where non-halting programs
 * accumulate void. The complement distribution then peaks at halting programs.
 */
export function programSpaceToVoidBoundary(ps: ProgramSpace): VoidBoundary {
  try {
    const boundary = createVoidBoundary(ps.totalPrograms);
    for (let i = 0; i < ps.programs.length; i++) {
      if (!ps.halting[i]) {
        updateVoidBoundary(boundary, i, 1);
      }
    }
    return boundary;
  } catch (error) {
    throw new Error(
      `[programSpaceToVoidBoundary] ${
        error instanceof Error ? error.message : String(error)
      }`
    );
  }
}

/**
 * The critical bridge: SolomonoffSpace to VoidBoundary.
 *
 * Rejection counts = complexity values. Higher complexity = more void
 * accumulated. complementDistribution() then peaks at the simplest
 * hypothesis (lowest complexity), recovering the Solomonoff prior ordering.
 *
 * This is the subsumption: Buleyean probability is void walking, with
 * complexity as the rejection predicate.
 */
export function solomonoffToVoidBoundary(ss: SolomonoffSpace): VoidBoundary {
  const n = ss.assignment.complexity.length;
  const boundary = createVoidBoundary(n);
  for (let i = 0; i < n; i++) {
    updateVoidBoundary(boundary, i, ss.assignment.complexity[i]);
  }
  return boundary;
}

/**
 * Buleyean weight for a choice in the Solomonoff space.
 * weight(i) = (ceiling + rounds + 1) - min(complexity[i], ceiling + rounds + 1) + 1
 *
 * Simpler hypotheses (lower complexity) get higher weight.
 * The +1 ensures positivity even at max complexity.
 */
export function solomonoffWeight(ss: SolomonoffSpace, idx: number): number {
  const cap = ss.assignment.ceiling + ss.rounds + 1;
  const c = Math.min(ss.assignment.complexity[idx], cap);
  return cap - c + 1;
}

// ============================================================================
// Solomonoff axiom verification
// ============================================================================

export interface SolomonoffAxiomResult {
  /** All weights are strictly positive */
  positivity: boolean;
  /** Weights can be normalized to sum to 1 */
  normalization: boolean;
  /** Simpler hypotheses have strictly higher weight than complex ones */
  concentration: boolean;
  /** All three axioms hold */
  allHold: boolean;
}

/**
 * Runtime verification of the three Solomonoff-Buleyean axioms:
 * 1. Positivity: every weight > 0
 * 2. Normalization: weights sum to a finite positive value
 * 3. Concentration: simpler hypotheses have higher weight
 */
export function verifySolomonoffAxioms(
  ss: SolomonoffSpace
): SolomonoffAxiomResult {
  try {
    const n = ss.assignment.complexity.length;
    if (n === 0) {
      return {
        positivity: true,
        normalization: true,
        concentration: true,
        allHold: true,
      };
    }

    const weights = Array.from({ length: n }, (_, i) =>
      solomonoffWeight(ss, i)
    );

    // Axiom 1: positivity
    const positivity = weights.every((w) => w > 0);

    // Axiom 2: normalization (weights sum to finite positive value)
    const sum = weights.reduce((a, b) => a + b, 0);
    const normalization = Number.isFinite(sum) && sum > 0;

    // Axiom 3: concentration
    // For any pair, lower complexity => higher weight
    let concentration = true;
    for (let i = 0; i < n && concentration; i++) {
      for (let j = i + 1; j < n && concentration; j++) {
        if (ss.assignment.complexity[i] < ss.assignment.complexity[j]) {
          if (weights[i] <= weights[j]) concentration = false;
        } else if (ss.assignment.complexity[i] > ss.assignment.complexity[j]) {
          if (weights[j] <= weights[i]) concentration = false;
        }
        // equal complexity => equal weight is fine
      }
    }

    return {
      positivity,
      normalization,
      concentration,
      allHold: positivity && normalization && concentration,
    };
  } catch (error) {
    throw new Error(
      `[verifySolomonoffAxioms] ${
        error instanceof Error ? error.message : String(error)
      }`
    );
  }
}

// ============================================================================
// Built-in classifiers
// ============================================================================

/**
 * Trivial classifier: even-length programs halt, odd-length don't.
 *
 * This is a minimal implementation for testing algorithmic invariants
 * without the computational overhead of actual program execution.
 * The even/odd rule provides a predictable halting distribution
 * that's useful for validating Omega approximation convergence.
 */
export function trivialClassifier(program: string): boolean {
  return program.length % 2 === 0;
}

/**
 * Step-bounded classifier with a 4-opcode instruction set.
 * Opcodes: HALT (0), INC (1), DEC (2), JNZ (3)
 *
 * This implements a minimal Turing-complete virtual machine for
 * realistic halting problem analysis. The 2-bit opcode encoding
 * allows compact binary program representation while providing
 * sufficient computational power for non-trivial behavior.
 *
 * The step limit prevents infinite loops while still allowing
 * observation of complex computational patterns. This mirrors
 * real-world constraints where we cannot wait indefinitely
 * for program termination.
 *
 * Interprets binary string as sequence of 2-bit opcodes.
 * Halts if HALT opcode is reached within stepLimit steps.
 */
export function stepBoundedClassifier(stepLimit: number): HaltingClassifier {
  return (program: string): boolean => {
    // Parse binary string into 2-bit opcodes
    const opcodes: number[] = [];
    for (let i = 0; i + 1 < program.length; i += 2) {
      opcodes.push(parseInt(program.slice(i, i + 2), 2));
    }
    if (opcodes.length === 0) return false;

    let pc = 0; // program counter
    let acc = 0; // accumulator
    let steps = 0;

    while (steps < stepLimit && pc < opcodes.length) {
      const op = opcodes[pc];
      switch (op) {
        case 0: // HALT
          return true;
        case 1: // INC
          acc++;
          pc++;
          break;
        case 2: // DEC
          acc = Math.max(0, acc - 1);
          pc++;
          break;
        case 3: // JNZ -- jump to start if acc != 0
          if (acc !== 0) {
            pc = 0;
          } else {
            pc++;
          }
          break;
      }
      steps++;
    }

    return false; // did not halt within step limit
  };
}

// ============================================================================
// Compression-based complexity estimation
// ============================================================================

/**
 * Estimate Kolmogorov complexity via compression length.
 * Uses run-length encoding as a simple compressor.
 *
 * This implements a practical approximation of Kolmogorov complexity,
 * which is theoretically uncomputable. Run-length encoding captures
 * regularity and repetition patterns that indicate low complexity.
 *
 * The invariance theorem guarantees that any compressor C provides
 * a valid upper bound: K(x) <= |C(x)| + O(1). While RLE is simple,
 * it's sufficient for distinguishing between random and structured
 * strings in many practical cases.
 *
 * Valid per the invariance theorem: K(x) <= |C(x)| + O(1) for any compressor C.
 */
export function estimateComplexity(input: string): number {
  try {
    if (input.length === 0) return 0;

    // Run-length encoding
    let compressed = '';
    let i = 0;
    while (i < input.length) {
      let run = 1;
      while (i + run < input.length && input[i + run] === input[i]) {
        run++;
      }
      compressed += run > 1 ? `${run}${input[i]}` : input[i];
      i += run;
    }

    return compressed.length;
  } catch (error) {
    throw new Error(
      `[estimateComplexity] ${
        error instanceof Error ? error.message : String(error)
      }`
    );
  }
}

/**
 * Build a ComplexityAssignment from a list of choice descriptions.
 * Each choice's complexity is estimated via compression.
 *
 * This function creates the data structure needed for Solomonoff
 * probability calculations. By estimating the Kolmogorov complexity
 * of each choice, we can assign prior probabilities that favor
 * simpler explanations - implementing Occam's razor algorithmically.
 *
 * The ceiling value represents the maximum complexity in the set,
 * which is used for normalization in the Buleyean weight formula.
 */
export function buildComplexityAssignment(
  choices: string[]
): ComplexityAssignment {
  const complexity = choices.map(estimateComplexity);
  const ceiling = complexity.length > 0 ? Math.max(...complexity) : 0;
  return { complexity, ceiling };
}

/**
 * Build a SolomonoffSpace from choice descriptions and empirical rounds.
 */
export function buildSolomonoffSpace(
  choices: string[],
  rounds: number = 0
): SolomonoffSpace {
  return {
    assignment: buildComplexityAssignment(choices),
    rounds,
    labels: choices,
  };
}
