// @a0n/.ignored_gnosis/src/chaitin-omega-handlers.ts
/**
 * Chaitin-Omega GG Topology Handlers
 *
 * Implements the node handlers for Chaitin's Omega and
 * Solomonoff-Buleyean subsumption as GnosisRegistry handlers.
 *
 * Follows the omega-checker-handlers.ts pattern exactly.
 */

import type { GnosisRegistry } from './runtime/registry';
import {
  type ProgramSpace,
  type SolomonoffSpace,
  type OmegaApproximation,
  type OmegaConvergenceTrace,
  type ComplexityAssignment,
  createProgramSpace,
  approximateOmega,
  trackOmegaConvergence,
  programSpaceToVoidBoundary,
  solomonoffToVoidBoundary,
  solomonoffWeight,
  verifySolomonoffAxioms,
  trivialClassifier,
  stepBoundedClassifier,
  estimateComplexity,
  buildComplexityAssignment,
} from './chaitin-omega';
import { complementDistribution } from './void';

// ─── Handlers ───────────────────────────────────────────────────────

/**
 * Handles the gnosis register Chaitin Omega Handlers workflow.
 */
export function registerChaitinOmegaHandlers(registry: GnosisRegistry): void {
  // Stage 1: Build ProgramSpace from props
  registry.register('ProgramSpaceEnumerator', async (_payload, props) => {
    const alphabetSize = parseInt(props.alphabetSize ?? '2', 10);
    const maxLength = parseInt(props.maxLength ?? '8', 10);
    const classifierType = props.classifier ?? 'trivial';
    const stepLimit = parseInt(props.stepLimit ?? '100', 10);

    const classifier =
      classifierType === 'step-bounded'
        ? stepBoundedClassifier(stepLimit)
        : trivialClassifier;

    const ps = createProgramSpace(alphabetSize, maxLength, classifier);
    return {
      alphabetSize: ps.alphabetSize,
      maxLength: ps.maxLength,
      totalPrograms: ps.totalPrograms,
      haltingPrograms: ps.haltingPrograms,
      nonHaltingPrograms: ps.nonHaltingPrograms,
      // Don't serialize the full program list for large spaces
      programCount: ps.programs.length,
      _programSpace: ps, // internal -- passed downstream
    };
  });

  // Stage 2: Compute OmegaApproximation from ProgramSpace
  registry.register('OmegaApproximator', async (payload) => {
    const ps = payload?._programSpace as ProgramSpace | undefined;
    if (!ps) {
      return { error: 'No ProgramSpace in payload' };
    }

    const omega = approximateOmega(ps);
    return {
      ...omega,
      _programSpace: ps, // pass through
    };
  });

  // Stage 3: Build SolomonoffSpace → VoidBoundary
  registry.register('SolomonoffInitializer', async (_payload, props) => {
    const choicesRaw = props.choices ?? '';
    const choices = choicesRaw
      .split(',')
      .map((s) => s.trim())
      .filter(Boolean);
    const rounds = parseInt(props.rounds ?? '0', 10);

    if (choices.length === 0) {
      return { error: 'No choices provided' };
    }

    const assignment = buildComplexityAssignment(choices);
    const ss: SolomonoffSpace = { assignment, rounds, labels: choices };
    const boundary = solomonoffToVoidBoundary(ss);
    const dist = complementDistribution(boundary);
    const axioms = verifySolomonoffAxioms(ss);

    return {
      choices,
      complexity: assignment.complexity,
      ceiling: assignment.ceiling,
      rounds,
      boundary: {
        counts: boundary.counts,
        totalEntries: boundary.totalEntries,
      },
      complementDistribution: dist,
      axioms,
      _solomonoffSpace: ss, // internal
    };
  });

  // Stage 4: Estimate K(x) via compression
  registry.register('ComplexityEstimator', async (payload, props) => {
    const input = props.input ?? (typeof payload === 'string' ? payload : '');
    const complexity = estimateComplexity(input);
    return {
      input,
      complexity,
      inputLength: input.length,
      compressionRatio: input.length > 0 ? complexity / input.length : 0,
    };
  });

  // Stage 5: Verify monotone convergence
  registry.register('OmegaConvergenceTracker', async (payload) => {
    const approximations = payload as OmegaApproximation[] | undefined;
    if (!Array.isArray(approximations)) {
      return { error: 'Expected array of OmegaApproximation' };
    }

    const trace = trackOmegaConvergence(approximations);
    return trace;
  });

  // Stage 6: Full bridge -- ProgramSpace → Omega + Solomonoff + verification
  registry.register('ChaitinSolomonoffBridge', async (_payload, props) => {
    const start = performance.now();
    const alphabetSize = parseInt(props.alphabetSize ?? '2', 10);
    const maxLength = parseInt(props.maxLength ?? '8', 10);
    const classifierType = props.classifier ?? 'trivial';
    const stepLimit = parseInt(props.stepLimit ?? '100', 10);
    const choicesRaw = props.choices ?? '';
    const choices = choicesRaw
      .split(',')
      .map((s) => s.trim())
      .filter(Boolean);
    const rounds = parseInt(props.rounds ?? '0', 10);

    const classifier =
      classifierType === 'step-bounded'
        ? stepBoundedClassifier(stepLimit)
        : trivialClassifier;

    // Build program space and omega
    const ps = createProgramSpace(alphabetSize, maxLength, classifier);
    const omega = approximateOmega(ps);
    const psBoundary = programSpaceToVoidBoundary(ps);

    // Build Solomonoff boundary
    let solomonoffResult = null;
    if (choices.length > 0) {
      const assignment = buildComplexityAssignment(choices);
      const ss: SolomonoffSpace = { assignment, rounds, labels: choices };
      const boundary = solomonoffToVoidBoundary(ss);
      const dist = complementDistribution(boundary);
      const axioms = verifySolomonoffAxioms(ss);
      const weights = choices.map((_, i) => solomonoffWeight(ss, i));

      solomonoffResult = {
        choices,
        complexity: assignment.complexity,
        ceiling: assignment.ceiling,
        weights,
        complementDistribution: dist,
        axioms,
      };
    }

    return {
      programSpace: {
        totalPrograms: ps.totalPrograms,
        haltingPrograms: ps.haltingPrograms,
        nonHaltingPrograms: ps.nonHaltingPrograms,
      },
      omega,
      psBoundary: {
        counts: psBoundary.counts,
        totalEntries: psBoundary.totalEntries,
      },
      solomonoff: solomonoffResult,
      verified:
        omega.ratio > 0 &&
        omega.ratio < 1 &&
        (solomonoffResult?.axioms.allHold ?? true),
      timeMs: performance.now() - start,
      status: 'VERIFIED',
    };
  });

  // Stage 7: Verify Solomonoff axioms standalone
  registry.register('SolomonoffAxiomChecker', async (payload) => {
    const ss = payload?._solomonoffSpace as SolomonoffSpace | undefined;
    if (!ss) {
      // Try to reconstruct from payload
      const choices = payload?.choices as string[] | undefined;
      const complexity = payload?.complexity as number[] | undefined;
      const ceiling = payload?.ceiling as number | undefined;
      const rounds = payload?.rounds as number | undefined;

      if (!complexity || !Array.isArray(complexity)) {
        return { error: 'No SolomonoffSpace or complexity array in payload' };
      }

      const reconstructed: SolomonoffSpace = {
        assignment: { complexity, ceiling: ceiling ?? Math.max(...complexity) },
        rounds: rounds ?? 0,
        labels: choices,
      };
      return verifySolomonoffAxioms(reconstructed);
    }

    return verifySolomonoffAxioms(ss);
  });
}

/** Theorem statements from ChaitinOmega.lean and SolomonoffBuleyean.lean */
export const CHAITIN_OMEGA_THEOREMS = [
  {
    name: 'omega_positivity',
    statement: '0 < Omega_L for all L >= 1 with halting programs',
  },
  {
    name: 'omega_subuniversality',
    statement: 'Omega_L < 1 for all finite L with non-halting programs',
  },
  {
    name: 'omega_monotone_convergence',
    statement: 'Omega_L <= Omega_{L+1} for all L',
  },
  {
    name: 'solomonoff_positivity',
    statement: 'weight(i) > 0 for all choices i',
  },
  {
    name: 'solomonoff_normalization',
    statement: 'sum(weight(i)) is finite and positive',
  },
  {
    name: 'solomonoff_concentration',
    statement: 'K(i) < K(j) implies weight(i) > weight(j)',
  },
  {
    name: 'solomonoff_buleyean_subsumption',
    statement:
      'Solomonoff prior = complementDistribution over complexity-weighted VoidBoundary',
  },
] as const;
