// @a0n/.ignored_gnosis/src/checkpoint-substrates-mcp.ts
/**
 * MCP Tool Definitions for Universal Checkpoint Reduction
 *
 * Three tools exposing the checkpoint substrate library to MCP consumers:
 *   - diagnose_substrate: map a substrate profile to CheckpointedSystem + diagnosis
 *   - forecast_repair: predict health after restoring k channels
 *   - compare_substrates: test topological equivalence of two profiles
 *
 * These tools are dependency-free (no zod). They validate inputs manually
 * and return MCP-compatible content arrays.
 *
 * @module @a0n/gnosis/checkpoint-substrates-mcp
 */

import type {
  CheckpointedSystem,
  Diagnosis,
  SubstrateComparison,
  SubstrateProfile,
} from './checkpoint-substrates';

import {
  substrateToUniversal,
  diagnose,
  forecastRepair,
  forecastNeglect,
  compareSubstrates,
} from './checkpoint-substrates';

// ═══════════════════════════════════════════════════════════════════════
// MCP-compatible result types (matches aeon-iam pattern)
// ═══════════════════════════════════════════════════════════════════════

export interface CheckpointMcpTextContent {
  type: 'text';
  text: string;
}

export interface CheckpointMcpToolResult {
  isError?: boolean;
  content: CheckpointMcpTextContent[];
}

export interface CheckpointMcpTool {
  name: string;
  description: string;
  inputSchema: Record<string, unknown>;
  execute: (args: Record<string, unknown>) => Promise<CheckpointMcpToolResult>;
}

// ═══════════════════════════════════════════════════════════════════════
// Helpers
// ═══════════════════════════════════════════════════════════════════════

function content(payload: unknown): CheckpointMcpToolResult {
  return {
    content: [
      {
        type: 'text',
        text: JSON.stringify(payload, null, 2),
      },
    ],
  };
}

function errorResult(message: string): CheckpointMcpToolResult {
  return {
    isError: true,
    content: [{ type: 'text', text: `Checkpoint MCP tool error: ${message}` }],
  };
}

const SUBSTRATE_TYPES = [
  'tokamak',
  'cardiac',
  'thermohaline',
  'turbulence',
  'codebase',
] as const;

type SubstrateType = (typeof SUBSTRATE_TYPES)[number];

function isSubstrateType(v: unknown): v is SubstrateType {
  return typeof v === 'string' && (SUBSTRATE_TYPES as readonly string[]).includes(v);
}

function parseNonNegInt(v: unknown, name: string): number {
  try {
    const n = typeof v === 'number' ? v : Number(v);
    if (!Number.isFinite(n) || n < 0 || !Number.isInteger(n)) {
      throw new Error(`${name} must be a non-negative integer, got ${v}`);
    }
    return n;

  } catch (error) {
    throw new Error(`[parseNonNegInt] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Parse a substrate profile from MCP tool arguments.
 */
function parseSubstrateProfile(args: Record<string, unknown>): SubstrateProfile {
  try {
    const type = args['substrate_type'] ?? args['substrateType'] ?? args['type'];
    if (!isSubstrateType(type)) {
      throw new Error(
        `substrate_type must be one of: ${SUBSTRATE_TYPES.join(', ')}; got ${type}`,
      );
    }

    const profile = (args['profile'] ?? args) as Record<string, unknown>;

    switch (type) {
      case 'tokamak':
        return {
          type: 'tokamak',
          profile: {
            activeModes: parseNonNegInt(
              profile['activeModes'] ?? profile['active_modes'],
              'activeModes',
            ),
            failedModes: parseNonNegInt(
              profile['failedModes'] ?? profile['failed_modes'],
              'failedModes',
            ),
          },
        };
      case 'cardiac':
        return {
          type: 'cardiac',
          profile: {
            activePathways: parseNonNegInt(
              profile['activePathways'] ?? profile['active_pathways'],
              'activePathways',
            ),
            failedPathways: parseNonNegInt(
              profile['failedPathways'] ?? profile['failed_pathways'],
              'failedPathways',
            ),
          },
        };
      case 'thermohaline':
        return {
          type: 'thermohaline',
          profile: {
            activeSites: parseNonNegInt(
              profile['activeSites'] ?? profile['active_sites'],
              'activeSites',
            ),
            disruptedSites: parseNonNegInt(
              profile['disruptedSites'] ?? profile['disrupted_sites'],
              'disruptedSites',
            ),
          },
        };
      case 'turbulence':
        return {
          type: 'turbulence',
          profile: {
            suppressedScales: parseNonNegInt(
              profile['suppressedScales'] ?? profile['suppressed_scales'],
              'suppressedScales',
            ),
            activeScales: parseNonNegInt(
              profile['activeScales'] ?? profile['active_scales'],
              'activeScales',
            ),
          },
        };
      case 'codebase':
        return {
          type: 'codebase',
          profile: {
            aliveConsumers: parseNonNegInt(
              profile['aliveConsumers'] ?? profile['alive_consumers'],
              'aliveConsumers',
            ),
            deadConsumers: parseNonNegInt(
              profile['deadConsumers'] ?? profile['dead_consumers'],
              'deadConsumers',
            ),
          },
        };
    }

  } catch (error) {
    throw new Error(`[parseSubstrateProfile] ${error instanceof Error ? error.message : String(error)}`);
  }
}

// ═══════════════════════════════════════════════════════════════════════
// JSON Schema fragments for MCP input schemas
// ═══════════════════════════════════════════════════════════════════════

const substrateTypeSchema = {
  type: 'string',
  enum: [...SUBSTRATE_TYPES],
  description:
    'The substrate domain: tokamak (plasma confinement), cardiac (electrophysiology), thermohaline (ocean circulation), turbulence (fluid dynamics), codebase (dependency topology).',
};

const profileSchema = {
  type: 'object',
  description:
    'Substrate-specific profile fields. For tokamak: activeModes, failedModes. For cardiac: activePathways, failedPathways. For thermohaline: activeSites, disruptedSites. For turbulence: suppressedScales, activeScales. For codebase: aliveConsumers, deadConsumers.',
  additionalProperties: true,
};

// ═══════════════════════════════════════════════════════════════════════
// Tool 1: diagnose_substrate
// ═══════════════════════════════════════════════════════════════════════

const diagnoseSubstrateTool: CheckpointMcpTool = {
  name: 'diagnose_substrate',
  description:
    'Map a substrate-specific profile (tokamak, cardiac, thermohaline, turbulence, codebase) to the universal CheckpointedSystem and return a health diagnosis. Returns beta1, deficit, sickness fraction, health status, repair path, and restoration/neglect forecasts.',
  inputSchema: {
    type: 'object',
    properties: {
      substrate_type: substrateTypeSchema,
      profile: profileSchema,
    },
    required: ['substrate_type', 'profile'],
  },
  execute: async (args) => {
    try {
      const sp = parseSubstrateProfile(args);
      const sys = substrateToUniversal(sp);
      const dx = diagnose(sys);
      return content({
        substrate: sp.type,
        system: sys,
        diagnosis: dx,
      });
    } catch (err) {
      return errorResult(err instanceof Error ? err.message : String(err));
    }
  },
};

// ═══════════════════════════════════════════════════════════════════════
// Tool 2: forecast_repair
// ═══════════════════════════════════════════════════════════════════════

const forecastRepairTool: CheckpointMcpTool = {
  name: 'forecast_repair',
  description:
    'Given a substrate profile and a proposed repair (k channels restored), forecast the new health state. Returns current system, repaired system, and diagnosis of both. Gain is always exactly k (linear, no diminishing returns). See codebase_adoption_gain, tokamak_coil_gain.',
  inputSchema: {
    type: 'object',
    properties: {
      substrate_type: substrateTypeSchema,
      profile: profileSchema,
      channels_restored: {
        type: 'integer',
        minimum: 0,
        description:
          'Number of channels to restore. Capped at the destroyed channel count.',
      },
    },
    required: ['substrate_type', 'profile', 'channels_restored'],
  },
  execute: async (args) => {
    try {
      const sp = parseSubstrateProfile(args);
      const k = parseNonNegInt(
        args['channels_restored'] ?? args['channelsRestored'],
        'channels_restored',
      );
      const current = substrateToUniversal(sp);
      const repaired = forecastRepair(current, k);
      const neglected = forecastNeglect(current, k);
      return content({
        substrate: sp.type,
        channelsRestored: Math.min(k, current.destroyed),
        current: { system: current, diagnosis: diagnose(current) },
        afterRepair: { system: repaired, diagnosis: diagnose(repaired) },
        afterNeglect: { system: neglected, diagnosis: diagnose(neglected) },
      });
    } catch (err) {
      return errorResult(err instanceof Error ? err.message : String(err));
    }
  },
};

// ═══════════════════════════════════════════════════════════════════════
// Tool 3: compare_substrates
// ═══════════════════════════════════════════════════════════════════════

const compareSubstratesTool: CheckpointMcpTool = {
  name: 'compare_substrates',
  description:
    'Compare two substrate profiles for topological equivalence. Two systems with the same deficit follow the same trajectory — all universal theorems apply identically. Returns deficit, sickness fraction, and which is healthier. See eight_substrates_preserve_deficit.',
  inputSchema: {
    type: 'object',
    properties: {
      a: {
        type: 'object',
        properties: {
          substrate_type: substrateTypeSchema,
          profile: profileSchema,
        },
        required: ['substrate_type', 'profile'],
        description: 'First substrate profile.',
      },
      b: {
        type: 'object',
        properties: {
          substrate_type: substrateTypeSchema,
          profile: profileSchema,
        },
        required: ['substrate_type', 'profile'],
        description: 'Second substrate profile.',
      },
    },
    required: ['a', 'b'],
  },
  execute: async (args) => {
    try {
      const aArgs = args['a'] as Record<string, unknown>;
      const bArgs = args['b'] as Record<string, unknown>;
      if (!aArgs || typeof aArgs !== 'object') {
        throw new Error('a must be an object with substrate_type and profile');
      }
      if (!bArgs || typeof bArgs !== 'object') {
        throw new Error('b must be an object with substrate_type and profile');
      }
      const spA = parseSubstrateProfile(aArgs);
      const spB = parseSubstrateProfile(bArgs);
      const sysA = substrateToUniversal(spA);
      const sysB = substrateToUniversal(spB);
      const comparison = compareSubstrates(sysA, sysB);
      return content({
        a: { substrate: spA.type, system: sysA, diagnosis: diagnose(sysA) },
        b: { substrate: spB.type, system: sysB, diagnosis: diagnose(sysB) },
        comparison,
      });
    } catch (err) {
      return errorResult(err instanceof Error ? err.message : String(err));
    }
  },
};

// ═══════════════════════════════════════════════════════════════════════
// Public API
// ═══════════════════════════════════════════════════════════════════════

/**
 * Create all checkpoint substrate MCP tools.
 *
 * Returns an array of tool definitions matching the MCP tool interface pattern.
 * No dependencies required — these tools are self-contained.
 *
 * Usage:
 * ```ts
 * import { createCheckpointMcpTools } from '@a0n/gnosis/checkpoint-substrates-mcp';
 * const tools = createCheckpointMcpTools();
 * // Register tools with your MCP server...
 * ```
 */
export function createCheckpointMcpTools(): CheckpointMcpTool[] {
  return [diagnoseSubstrateTool, forecastRepairTool, compareSubstratesTool];
}

export {
  type CheckpointedSystem,
  type Diagnosis,
  type SubstrateComparison,
  type SubstrateProfile,
};
