// @a0n/.ignored_gnosis/src/checkpoint-substrates.ts
/**
 * Universal Checkpoint Reduction — TypeScript Substrate Mappings
 *
 * Mirrors the Lean proofs in:
 *   - ForkRaceFoldTheorems/UniversalCheckpointReduction.lean
 *   - ForkRaceFoldTheorems/UniversalCheckpointSubstrates.lean
 *   - ForkRaceFoldTheorems/CodebaseCheckpointSubstrate.lean
 *
 * Every system that satisfies three axioms exhibits the same topological
 * phase transitions — cancer, trading, weather, turbulence, power grids,
 * plasma confinement, cardiac rhythm, ocean circulation, codebases.
 *
 * The three axioms:
 *   1. Finite rejection channels
 *   2. beta1 as rejection capacity (additive)
 *   3. Clinamen cost: +1 perturbation per channel change
 *
 * @module @a0n/gnosis/checkpoint-substrates
 */

// ═══════════════════════════════════════════════════════════════════════
// The Universal System
// ═══════════════════════════════════════════════════════════════════════

/**
 * A substrate-independent checkpointed system.
 *
 * Maps to Lean's `CheckpointedSystem` in UniversalCheckpointReduction.lean.
 * All substrate-specific profiles reduce to this structure.
 */
export interface CheckpointedSystem {
  /** Active (functional) rejection channel count. */
  active: number;
  /** Destroyed (knocked-out) rejection channel count. */
  destroyed: number;
  /**
   * Total rejection capacity of active channels.
   * Mirrors `CheckpointedSystem.beta1` in the Lean proof.
   */
  beta1: number;
  /**
   * Topological deficit: how many rejection paths have been lost.
   * Mirrors `CheckpointedSystem.deficit` in the Lean proof.
   */
  deficit: number;
  /**
   * Total capacity if all channels were active (healthy state).
   * Mirrors `CheckpointedSystem.healthyBeta1` in the Lean proof.
   * Invariant: healthyBeta1 === beta1 + deficit (deficit_is_signature).
   */
  healthyBeta1: number;
  /**
   * The system is deaf when all channels are destroyed.
   * Mirrors `CheckpointedSystem.isDeaf` — `active === []`.
   * @see universal_deaf_is_zero
   */
  isDeaf: boolean;
  /**
   * The system has positive rejection capacity.
   * Mirrors `CheckpointedSystem.isLearning` — `0 < beta1`.
   */
  isLearning: boolean;
  /**
   * Sickness fraction: deficit / (beta1 + deficit), in [0..1].
   * When 0 the organism is fully healthy; when 1 it is fully cancerous.
   * @see codebase_sickness_boundary_healthy, codebase_sickness_boundary_cancerous
   */
  sicknessFraction: number;
}

// ═══════════════════════════════════════════════════════════════════════
// Substrate Profiles
// ═══════════════════════════════════════════════════════════════════════

/**
 * Tokamak plasma confinement profile.
 * Rejection channels = magnetic confinement modes.
 * @see TokamakProfile in UniversalCheckpointSubstrates.lean
 */
export interface TokamakProfile {
  activeModes: number;
  failedModes: number;
}

/**
 * Cardiac electrophysiology profile.
 * Rejection channels = conduction pathways (SA node, AV node, His, Purkinje).
 * @see CardiacProfile in UniversalCheckpointSubstrates.lean
 */
export interface CardiacProfile {
  activePathways: number;
  failedPathways: number;
}

/**
 * Thermohaline circulation profile.
 * Rejection channels = deep-water formation sites.
 * @see ThermohalineProfile in UniversalCheckpointSubstrates.lean
 */
export interface ThermohalineProfile {
  activeSites: number;
  disruptedSites: number;
}

/**
 * Turbulent flow profile.
 * Rejection channels = eddy cascade levels that suppress turbulence.
 * Note: in the Lean proof, `activeScales` counts destroyed channels
 * (scales that have gone turbulent), consistent with the turbulence
 * substrate where active turbulence = lost suppressive capacity.
 * @see TurbulenceProfile in UniversalCheckpointSubstrates.lean
 */
export interface TurbulenceProfile {
  suppressedScales: number;
  activeScales: number;
}

/**
 * Codebase dependency topology profile.
 * Rejection channels = reverse dependencies (alive consumers).
 * @see CodebaseProfile in CodebaseCheckpointSubstrate.lean
 */
export interface CodebaseProfile {
  aliveConsumers: number;
  deadConsumers: number;
}

/**
 * Union type of all supported substrate profiles.
 */
export type SubstrateProfile =
  | { type: 'tokamak'; profile: TokamakProfile }
  | { type: 'cardiac'; profile: CardiacProfile }
  | { type: 'thermohaline'; profile: ThermohalineProfile }
  | { type: 'turbulence'; profile: TurbulenceProfile }
  | { type: 'codebase'; profile: CodebaseProfile };

// ═══════════════════════════════════════════════════════════════════════
// Internal: build a CheckpointedSystem from (active, destroyed) counts
// ═══════════════════════════════════════════════════════════════════════

function buildSystem(active: number, destroyed: number): CheckpointedSystem {
  const beta1 = active;
  const deficit = destroyed;
  const healthyBeta1 = beta1 + deficit;
  const total = beta1 + deficit;
  return {
    active,
    destroyed,
    beta1,
    deficit,
    healthyBeta1,
    isDeaf: active === 0,
    isLearning: active > 0,
    sicknessFraction: total > 0 ? deficit / total : 0,
  };
}

// ═══════════════════════════════════════════════════════════════════════
// Substrate Mapping Functions
// ═══════════════════════════════════════════════════════════════════════

/**
 * Map a tokamak confinement profile to the universal checkpointed system.
 *
 * Mirrors `tokamakToUniversal` in UniversalCheckpointSubstrates.lean.
 * Preserves deficit: `tokamak_maps_to_universal`.
 *
 * @see tokamak_disruption_is_collapse — all modes failed => beta1 = 0
 * @see tokamak_coil_restoration_restores — +1 mode => beta1 > 0
 * @see tokamak_conservation — beta1 + deficit = activeModes + failedModes
 */
export function tokamakToUniversal(tp: TokamakProfile): CheckpointedSystem {
  return buildSystem(tp.activeModes, tp.failedModes);
}

/**
 * Map a cardiac electrophysiology profile to the universal checkpointed system.
 *
 * Mirrors `cardiacToUniversal` in UniversalCheckpointSubstrates.lean.
 * Preserves deficit: `cardiac_maps_to_universal`.
 *
 * @see fibrillation_is_collapse — all pathways failed => beta1 = 0
 * @see defibrillation_is_sovereign_restoration — +1 pathway => beta1 > 0
 */
export function cardiacToUniversal(cp: CardiacProfile): CheckpointedSystem {
  return buildSystem(cp.activePathways, cp.failedPathways);
}

/**
 * Map a thermohaline circulation profile to the universal checkpointed system.
 *
 * Mirrors `thermohalineToUniversal` in UniversalCheckpointSubstrates.lean.
 * Preserves deficit: `thermohaline_maps_to_universal`.
 *
 * @see amoc_shutdown_is_collapse — all sites disrupted => beta1 = 0
 */
export function thermohalineToUniversal(
  tp: ThermohalineProfile,
): CheckpointedSystem {
  return buildSystem(tp.activeSites, tp.disruptedSites);
}

/**
 * Map a turbulence profile to the universal checkpointed system.
 *
 * Mirrors `turbulenceToUniversal` in UniversalCheckpointSubstrates.lean.
 * Preserves deficit: `turbulence_maps_to_universal`.
 *
 * @see full_turbulence_is_collapse — all suppression lost => beta1 = 0
 */
export function turbulenceToUniversal(
  tp: TurbulenceProfile,
): CheckpointedSystem {
  return buildSystem(tp.suppressedScales, tp.activeScales);
}

/**
 * Map a codebase dependency profile to the universal checkpointed system.
 *
 * Mirrors `codebaseToUniversal` in CodebaseCheckpointSubstrate.lean.
 * Preserves deficit: `codebase_maps_to_universal`.
 *
 * @see orphan_is_collapse — zero alive consumers => beta1 = 0
 * @see adoption_restores — +1 consumer => beta1 > 0
 * @see codebase_conservation — beta1 + deficit = alive + dead
 * @see codebase_adoption_gain — adopting k orphans increases beta1 by k
 */
export function codebaseToUniversal(cp: CodebaseProfile): CheckpointedSystem {
  return buildSystem(cp.aliveConsumers, cp.deadConsumers);
}

/**
 * Convert any tagged substrate profile to a universal checkpointed system.
 */
export function substrateToUniversal(sp: SubstrateProfile): CheckpointedSystem {
  try {
    switch (sp.type) {
      case 'tokamak':
        return tokamakToUniversal(sp.profile);
      case 'cardiac':
        return cardiacToUniversal(sp.profile);
      case 'thermohaline':
        return thermohalineToUniversal(sp.profile);
      case 'turbulence':
        return turbulenceToUniversal(sp.profile);
      case 'codebase':
        return codebaseToUniversal(sp.profile);
    }

  } catch (error) {
    throw new Error(`[substrateToUniversal] ${error instanceof Error ? error.message : String(error)}`);
  }
}

// ═══════════════════════════════════════════════════════════════════════
// Diagnostic Functions
// ═══════════════════════════════════════════════════════════════════════

/** Health classification based on sickness fraction thresholds. */
export type HealthStatus =
  | 'healthy'
  | 'stressed'
  | 'inflamed'
  | 'cancerous'
  | 'necrotic';

/** Repair path following the GallWaspTestimony. */
export type RepairPath = 'integration' | 'exorcism' | 'monitoring';

/**
 * Diagnosis result for a checkpointed system.
 */
export interface Diagnosis {
  /** Overall health classification. */
  health: HealthStatus;
  /**
   * Recommended repair path.
   * - monitoring: sickness < 0.2, system is mostly healthy
   * - integration: sickness 0.2..0.8, adopt orphans to restore beta1
   * - exorcism: sickness >= 0.8, remove the cancer entirely
   */
  repairPath: RepairPath;
  /**
   * How much beta1 increases if 1 channel is restored.
   * Always 1 (by the clinamen cost axiom).
   * @see universal_one_channel_restores
   */
  restorationGain: number;
  /**
   * Predicted beta1 after losing 1 more channel.
   * @see codebase_forecast_deficit_rate, tokamak_forecast_disruption_rate
   */
  forecastIfNeglected: number;
}

/**
 * Diagnose a checkpointed system's health and recommend a repair path.
 *
 * Health thresholds (sickness fraction):
 *   - healthy:   [0, 0.1)
 *   - stressed:  [0.1, 0.3)
 *   - inflamed:  [0.3, 0.6)
 *   - cancerous: [0.6, 1.0)
 *   - necrotic:  1.0 (fully deaf)
 *
 * Repair paths:
 *   - monitoring:   sickness < 0.2
 *   - integration:  0.2 <= sickness < 0.8
 *   - exorcism:     sickness >= 0.8
 *
 * @see universal_one_channel_restores — restorationGain is always 1
 * @see codebase_forecast_deficit_rate — forecastIfNeglected is exact
 */
export function diagnose(sys: CheckpointedSystem): Diagnosis {
  try {
    const sf = sys.sicknessFraction;

    let health: HealthStatus;
    if (sf >= 1.0 || sys.isDeaf) {
      health = 'necrotic';
    } else if (sf >= 0.6) {
      health = 'cancerous';
    } else if (sf >= 0.3) {
      health = 'inflamed';
    } else if (sf >= 0.1) {
      health = 'stressed';
    } else {
      health = 'healthy';
    }

    let repairPath: RepairPath;
    if (sf >= 0.8) {
      repairPath = 'exorcism';
    } else if (sf >= 0.2) {
      repairPath = 'integration';
    } else {
      repairPath = 'monitoring';
    }

    return {
      health,
      repairPath,
      // By the clinamen cost axiom, restoring 1 channel always adds 1 to beta1
      restorationGain: 1,
      // By the forecast theorem, losing 1 channel drops beta1 by exactly 1
      forecastIfNeglected: Math.max(0, sys.beta1 - 1),
    };

  } catch (error) {
    throw new Error(`[diagnose] ${error instanceof Error ? error.message : String(error)}`);
  }
}

// ═══════════════════════════════════════════════════════════════════════
// God Formula: Universal Restoration Priority
// ═══════════════════════════════════════════════════════════════════════

/**
 * Compute the God Formula restoration weight for a position in the void.
 *
 *   w(R, v) = R - min(v, R) + 1
 *
 * Where R = total observation rounds, v = rejection count (destroyed channels).
 *
 * - Sovereign positions (v=0): weight = R + 1 (highest priority)
 * - Oracle positions (v >= R): weight = 1 (lowest priority)
 *
 * Mirrors `godWeight` in BuleyeanProbability.lean / UniversalCheckpointReduction.lean.
 *
 * @see universal_sovereign_highest — sovereign always has highest weight
 * @see universal_restoration_positive — weight is always > 0
 */
export function godFormulaWeight(totalRounds: number, rejections: number): number {
  return totalRounds - Math.min(rejections, totalRounds) + 1;
}

/**
 * Compute restoration weight for a checkpointed system.
 *
 * Uses the system's healthyBeta1 as R (total rounds = total channels)
 * and the system's deficit as v (rejections = destroyed channels).
 *
 * @see UniversalRestorationCandidate.restorationValue in UniversalCheckpointReduction.lean
 */
export function systemRestorationWeight(sys: CheckpointedSystem): number {
  return godFormulaWeight(sys.healthyBeta1, sys.deficit);
}

// ═══════════════════════════════════════════════════════════════════════
// Forecasting and Comparison
// ═══════════════════════════════════════════════════════════════════════

/**
 * Forecast the system state after restoring k channels.
 *
 * @see codebase_adoption_gain — adopting k orphans increases beta1 by exactly k
 * @see tokamak_coil_gain — restoring k coils increases beta1 by exactly k
 */
export function forecastRepair(
  sys: CheckpointedSystem,
  channelsRestored: number,
): CheckpointedSystem {
  const k = Math.min(channelsRestored, sys.destroyed);
  return buildSystem(sys.active + k, sys.destroyed - k);
}

/**
 * Forecast the system state after losing k more channels.
 *
 * @see codebase_forecast_deficit_rate — losing d channels drops beta1 by d
 * @see tokamak_forecast_disruption_rate — losing d modes drops beta1 by d
 */
export function forecastNeglect(
  sys: CheckpointedSystem,
  channelsLost: number,
): CheckpointedSystem {
  const d = Math.min(channelsLost, sys.active);
  return buildSystem(sys.active - d, sys.destroyed + d);
}

/**
 * Compare two substrate profiles for topological equivalence.
 *
 * Two systems are topologically equivalent if they have the same deficit.
 * Same deficit => same trajectory => same universal theorems apply identically.
 *
 * @see eight_substrates_preserve_deficit — all substrates preserve deficit
 */
export function areTopologicallyEquivalent(
  a: CheckpointedSystem,
  b: CheckpointedSystem,
): boolean {
  return a.deficit === b.deficit;
}

/**
 * Compare two substrate profiles and describe their relationship.
 */
export interface SubstrateComparison {
  /** Whether the two systems have the same deficit (same trajectory). */
  topologicallyEquivalent: boolean;
  /** Deficit of system A. */
  deficitA: number;
  /** Deficit of system B. */
  deficitB: number;
  /** Sickness fraction of system A. */
  sicknessFractionA: number;
  /** Sickness fraction of system B. */
  sicknessFractionB: number;
  /** Which system is healthier, or 'equal'. */
  healthier: 'a' | 'b' | 'equal';
}

/**
 * Compare two checkpointed systems.
 */
export function compareSubstrates(
  a: CheckpointedSystem,
  b: CheckpointedSystem,
): SubstrateComparison {
  return {
    topologicallyEquivalent: a.deficit === b.deficit,
    deficitA: a.deficit,
    deficitB: b.deficit,
    sicknessFractionA: a.sicknessFraction,
    sicknessFractionB: b.sicknessFraction,
    healthier:
      a.sicknessFraction < b.sicknessFraction
        ? 'a'
        : a.sicknessFraction > b.sicknessFraction
          ? 'b'
          : 'equal',
  };
}
