// @a0n/.ignored_gnosis/src/clock-atlas.ts
/**
 * Clock Atlas -- convert a DistributedClockMesh into visualization-ready data.
 *
 * Produces a ClockAtlasPacket suitable for rendering on a world map:
 * bodies (nodes with geographic coordinates) and edges (propagation
 * links with delay and causal classification).
 *
 * @packageDocumentation
 */

import type {
  DistributedClockMesh,
  ProcessPhase,
} from './clock-distributed';
import { coloToLatLon } from './colo-coordinates';

// ─── Atlas Types ────────────────────────────────────────────────────

/** A positioned node in the clock atlas. */
export interface ClockAtlasBody {
  readonly id: string;
  readonly label: string;
  readonly kind: 'aeonpid';
  readonly colo: string;
  readonly city: string;
  readonly phase: ProcessPhase;
  readonly lat: number;
  readonly lon: number;
  readonly requestRate: number;
  readonly clockPLo: number;
}

/** A propagation link between two atlas bodies. */
export interface ClockAtlasEdge {
  readonly fromId: string;
  readonly toId: string;
  readonly propagationMs: number;
  readonly edgeType: 'PROCESS' | 'VENT';
}

/** Complete atlas packet for visualization. */
export interface ClockAtlasPacket {
  readonly source: 'GnosisClockAtlas';
  readonly bodies: readonly ClockAtlasBody[];
  readonly edges: readonly ClockAtlasEdge[];
  readonly networkDiameterMs: number;
  readonly litFraction: number;
  readonly activeCount: number;
  readonly totalCount: number;
}

// ─── Builder ────────────────────────────────────────────────────────

/**
 * Convert a DistributedClockMesh into a ClockAtlasPacket.
 *
 * Each ProcessClock maps to a body with geographic coordinates
 * looked up from its region (treated as a colo code). Each mesh
 * edge maps to an atlas edge with propagation delay and type.
 */
export function createClockAtlas(mesh: DistributedClockMesh): ClockAtlasPacket {
  const bodies: ClockAtlasBody[] = [];

  for (const clock of mesh.clocks) {
    // The region field stores the colo code (set by clock-topology-coordinator)
    const coords = coloToLatLon(clock.location.region);
    const lat = coords?.lat ?? clock.location.lat;
    const lon = coords?.lon ?? clock.location.lon;
    const city = coords?.city ?? clock.location.region;

    bodies.push({
      id: clock.id,
      label: clock.label,
      kind: 'aeonpid',
      colo: clock.location.region,
      city,
      phase: clock.phase,
      lat,
      lon,
      requestRate: clock.requestRate,
      clockPLo: clock.localEpochMs,
    });
  }

  const edges: ClockAtlasEdge[] = mesh.edges.map(e => ({
    fromId: e.fromId,
    toId: e.toId,
    propagationMs: e.distance.propagationDelayMs,
    edgeType: e.edgeType,
  }));

  return {
    source: 'GnosisClockAtlas',
    bodies,
    edges,
    networkDiameterMs: mesh.networkDiameterMs,
    litFraction: mesh.litFraction,
    activeCount: mesh.activeCount,
    totalCount: mesh.clocks.length,
  };
}
