// @a0n/.ignored_gnosis/src/cross-domain.test.ts
import { describe, expect, it } from '@a0n/gnosis/test';
import {
  LanguageAdapterRegistry,
  crossCompile,
  type CrossDomainAnalysis,
  type SemanticContract,
} from './lib';

const EMPTY_CONTRACT: SemanticContract = {
  paramTypes: [],
  returnType: { kind: 'unknown' },
  predicates: [],
  facets: [],
  obligations: [],
};

describe('cross-domain preservation', () => {
  it('fails closed when a required nuance facet is dropped', async () => {
    const registry = new LanguageAdapterRegistry();

    registry.register({
      id: 'test:natural-source',
      domain: 'natural',
      languages: ['en'],
      analyze(source): CrossDomainAnalysis {
        return {
          adapterId: 'test:natural-source',
          source,
          semanticContract: EMPTY_CONTRACT,
          facets: [
            {
              key: 'tone',
              value: 'polite',
              domain: 'natural',
              required: true,
            },
          ],
          obligations: [],
          diagnostics: [],
        };
      },
    });

    registry.register({
      id: 'test:code-target',
      domain: 'code',
      languages: ['typescript'],
      emit(_analysis, target) {
        return {
          adapterId: 'test:code-target',
          domain: 'code',
          language: target.language,
          text: 'export function translated() {}',
          facets: [],
          obligations: [],
          diagnostics: [],
        };
      },
    });

    const result = await crossCompile({
      source: {
        domain: 'natural',
        language: 'en',
        sourceText: 'Please help me.',
      },
      target: {
        domain: 'code',
        language: 'typescript',
      },
      preservation: {
        required: ['tone'],
      },
      registry,
    });

    expect(result.success).toBe(false);
    expect(
      result.obligations.some(
        (obligation) =>
          obligation.key === 'preserve-tone' && obligation.severity === 'error'
      )
    ).toBe(true);
  });
});
