// @a0n/.ignored_gnosis/src/css-cascade-topology.ts
/**
 * CSS Cascade Topology -- Deterministic style resolution over known trees
 *
 * The cascade is not magic. It is a directed acyclic graph:
 *   - FORK: one CSS rule fans out to every node it matches
 *   - RACE: multiple rules compete for the same property on the same node
 *   - FOLD: specificity sorts competitors; winner takes the property
 *   - TRACE: inheritance feeds resolved parent styles into child resolution
 *
 * If you know the tree and you know the CSS, the resolved styles at every
 * leaf are a deterministic function. This module computes that function.
 *
 * @module @a0n/gnosis/css-cascade-topology
 */

// ── Tree types (compatible with aeon-flux TreeNode / SerializedComponent) ────

export interface CascadeNode {
  type: string;
  props?: Record<string, unknown>;
  children?: (CascadeNode | string)[];
}

export interface ResolvedNode {
  type: string;
  props?: Record<string, unknown>;
  children?: (ResolvedNode | string)[];
  /** The final computed styles for this node */
  resolvedStyles: Record<string, string>;
  /** Every rule that matched, with its specificity -- for diagnostics */
  matchedRules: MatchedRule[];
  /** Properties where two+ rules had equal specificity (cascade ambiguity) */
  conflicts: PropertyConflict[];
}

export interface MatchedRule {
  selector: string;
  specificity: Specificity;
  declarations: Declaration[];
  sourceOrder: number;
}

export interface PropertyConflict {
  property: string;
  contenders: { selector: string; specificity: Specificity; value: string }[];
  winner: string;
  reason: 'specificity' | 'source-order';
}

// ── CSS types ────────────────────────────────────────────────────────────────

export interface CSSRule {
  selector: string;
  declarations: Declaration[];
  sourceOrder: number;
}

export interface Declaration {
  property: string;
  value: string;
  important: boolean;
}

/**
 * Specificity as (a, b, c):
 *   a = ID selectors
 *   b = class selectors + attribute selectors + pseudo-classes
 *   c = type selectors + pseudo-elements
 */
export type Specificity = [number, number, number];

// ── Parsed selector types ────────────────────────────────────────────────────

type SimpleSelectorType = 'type' | 'class' | 'id' | 'universal' | 'attribute' | 'pseudo-class' | 'pseudo-element';

interface SimpleSelector {
  kind: SimpleSelectorType;
  value: string;
  /** For attribute selectors: [attr=value] */
  attr?: string;
  attrOp?: string;
  attrValue?: string;
}

type Combinator = 'descendant' | 'child' | 'adjacent' | 'general-sibling';

interface CompoundSelector {
  simples: SimpleSelector[];
}

interface SelectorSequence {
  compounds: CompoundSelector[];
  combinators: Combinator[];
}

// ── CSS Parser ───────────────────────────────────────────────────────────────

/**
 * Parse a CSS string into rules. Not a full CSS parser -- handles the
 * subset that actually appears in component stylesheets: selectors,
 * declarations, !important, @media (ignored), and comments.
 */
export function parseCSS(css: string): CSSRule[] {
  try {
    const rules: CSSRule[] = [];
    let order = 0;

    // Strip comments
    const stripped = css.replace(/\/\*[\s\S]*?\*\//g, '');

    // Simple block-level tokenizer
    let i = 0;
    while (i < stripped.length) {
      // Skip whitespace
      while (i < stripped.length && /\s/.test(stripped[i]!)) i++;
      if (i >= stripped.length) break;

      // Skip @-rules (media, keyframes, etc.) -- find matching brace depth
      if (stripped[i] === '@') {
        const atStart = i;
        // Find the opening brace or semicolon
        while (i < stripped.length && stripped[i] !== '{' && stripped[i] !== ';') i++;
        if (i >= stripped.length) break;
        if (stripped[i] === ';') {
          i++;
          continue;
        }
        // @media: recurse into the block
        if (stripped.slice(atStart, i).trim().startsWith('@media')) {
          i++; // skip {
          let depth = 1;
          const mediaStart = i;
          while (i < stripped.length && depth > 0) {
            if (stripped[i] === '{') depth++;
            else if (stripped[i] === '}') depth--;
            i++;
          }
          // Parse the inner rules
          const inner = stripped.slice(mediaStart, i - 1);
          const innerRules = parseCSS(inner);
          for (const r of innerRules) {
            r.sourceOrder = order++;
            rules.push(r);
          }
          continue;
        }
        // Other @-rules: skip the block
        i++; // skip {
        let depth = 1;
        while (i < stripped.length && depth > 0) {
          if (stripped[i] === '{') depth++;
          else if (stripped[i] === '}') depth--;
          i++;
        }
        continue;
      }

      // Read selector (everything before {)
      const selStart = i;
      while (i < stripped.length && stripped[i] !== '{') i++;
      if (i >= stripped.length) break;
      const selectorText = stripped.slice(selStart, i).trim();
      i++; // skip {

      // Read declarations (everything before })
      const declStart = i;
      while (i < stripped.length && stripped[i] !== '}') i++;
      const declText = stripped.slice(declStart, i).trim();
      i++; // skip }

      if (!selectorText || !declText) continue;

      const declarations = parseDeclarations(declText);

      // Handle comma-separated selectors
      const selectors = selectorText.split(',').map(s => s.trim()).filter(Boolean);
      for (const sel of selectors) {
        rules.push({ selector: sel, declarations, sourceOrder: order++ });
      }
    }

    return rules;

  } catch (error) {
    throw new Error(`[parseCSS] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function parseDeclarations(text: string): Declaration[] {
  try {
    const decls: Declaration[] = [];
    // Split on semicolons, but respect parentheses (for values like rgb(), var())
    const parts: string[] = [];
    let current = '';
    let parenDepth = 0;
    for (const ch of text) {
      if (ch === '(' ) parenDepth++;
      else if (ch === ')') parenDepth--;
      else if (ch === ';' && parenDepth === 0) {
        parts.push(current.trim());
        current = '';
        continue;
      }
      current += ch;
    }
    if (current.trim()) parts.push(current.trim());

    for (const part of parts) {
      const colonIdx = part.indexOf(':');
      if (colonIdx === -1) continue;
      const property = part.slice(0, colonIdx).trim();
      let value = part.slice(colonIdx + 1).trim();
      const important = value.endsWith('!important') || value.includes('!important');
      if (important) {
        value = value.replace(/\s*!important\s*$/, '').trim();
      }
      if (property && value) {
        decls.push({ property, value, important });
      }
    }
    return decls;

  } catch (error) {
    throw new Error(`[parseDeclarations] ${error instanceof Error ? error.message : String(error)}`);
  }
}

// ── Selector Parser ──────────────────────────────────────────────────────────

function parseSelectorSequence(selector: string): SelectorSequence {
  try {
    const compounds: CompoundSelector[] = [];
    const combinators: Combinator[] = [];

    const tokens = tokenizeSelector(selector);
    let current: SimpleSelector[] = [];

    for (const tok of tokens) {
      if (tok.type === 'combinator') {
        if (current.length > 0) {
          compounds.push({ simples: current });
          current = [];
        }
        combinators.push(tok.value as Combinator);
      } else {
        current.push(tok.simple!);
      }
    }
    if (current.length > 0) {
      compounds.push({ simples: current });
    }

    return { compounds, combinators };

  } catch (error) {
    throw new Error(`[parseSelectorSequence] ${error instanceof Error ? error.message : String(error)}`);
  }
}

interface SelectorToken {
  type: 'simple' | 'combinator';
  simple?: SimpleSelector;
  value?: string;
}

function tokenizeSelector(sel: string): SelectorToken[] {
  try {
    const tokens: SelectorToken[] = [];
    let i = 0;
    const s = sel.trim();

    while (i < s.length) {
      // Skip whitespace -- might be descendant combinator
      if (/\s/.test(s[i]!)) {
        while (i < s.length && /\s/.test(s[i]!)) i++;
        if (i >= s.length) break;
        // Check if next char is an explicit combinator
        if (s[i] === '>' || s[i] === '+' || s[i] === '~') continue;
        // Otherwise it's a descendant combinator
        tokens.push({ type: 'combinator', value: 'descendant' });
        continue;
      }

      // Explicit combinators
      if (s[i] === '>') {
        i++;
        while (i < s.length && /\s/.test(s[i]!)) i++;
        tokens.push({ type: 'combinator', value: 'child' });
        continue;
      }
      if (s[i] === '+') {
        i++;
        while (i < s.length && /\s/.test(s[i]!)) i++;
        tokens.push({ type: 'combinator', value: 'adjacent' });
        continue;
      }
      if (s[i] === '~') {
        i++;
        while (i < s.length && /\s/.test(s[i]!)) i++;
        tokens.push({ type: 'combinator', value: 'general-sibling' });
        continue;
      }

      // ID selector
      if (s[i] === '#') {
        i++;
        const start = i;
        while (i < s.length && /[-\w]/.test(s[i]!)) i++;
        tokens.push({ type: 'simple', simple: { kind: 'id', value: s.slice(start, i) } });
        continue;
      }

      // Class selector
      if (s[i] === '.') {
        i++;
        const start = i;
        while (i < s.length && /[-\w]/.test(s[i]!)) i++;
        tokens.push({ type: 'simple', simple: { kind: 'class', value: s.slice(start, i) } });
        continue;
      }

      // Universal selector
      if (s[i] === '*') {
        i++;
        tokens.push({ type: 'simple', simple: { kind: 'universal', value: '*' } });
        continue;
      }

      // Attribute selector
      if (s[i] === '[') {
        i++;
        const start = i;
        while (i < s.length && s[i] !== ']') i++;
        const inner = s.slice(start, i);
        i++; // skip ]
        const attrMatch = inner.match(/^([\w-]+)(?:([~|^$*]?=)"?([^"]*)"?)?$/);
        if (attrMatch) {
          tokens.push({
            type: 'simple',
            simple: {
              kind: 'attribute',
              value: inner,
              attr: attrMatch[1],
              attrOp: attrMatch[2],
              attrValue: attrMatch[3],
            },
          });
        }
        continue;
      }

      // Pseudo-class or pseudo-element
      if (s[i] === ':') {
        i++;
        const isPseudoElement = s[i] === ':';
        if (isPseudoElement) i++;
        const start = i;
        while (i < s.length && /[-\w]/.test(s[i]!)) i++;
        let value = s.slice(start, i);
        // Handle functional pseudo-classes like :not(), :is(), :where()
        if (i < s.length && s[i] === '(') {
          i++; // skip (
          let depth = 1;
          const argStart = i;
          while (i < s.length && depth > 0) {
            if (s[i] === '(') depth++;
            else if (s[i] === ')') depth--;
            i++;
          }
          value += '(' + s.slice(argStart, i - 1) + ')';
        }
        tokens.push({
          type: 'simple',
          simple: {
            kind: isPseudoElement ? 'pseudo-element' : 'pseudo-class',
            value,
          },
        });
        continue;
      }

      // Type selector
      if (/[-\w]/.test(s[i]!)) {
        const start = i;
        while (i < s.length && /[-\w]/.test(s[i]!)) i++;
        tokens.push({ type: 'simple', simple: { kind: 'type', value: s.slice(start, i) } });
        continue;
      }

      // Unknown character -- skip
      i++;
    }

    return tokens;

  } catch (error) {
    throw new Error(`[tokenizeSelector] ${error instanceof Error ? error.message : String(error)}`);
  }
}

// ── Specificity Calculator ───────────────────────────────────────────────────

/**
 * Computes the Specificity.
 */
export function computeSpecificity(selector: string): Specificity {
  try {
    const seq = parseSelectorSequence(selector);
    let a = 0, b = 0, c = 0;

    for (const compound of seq.compounds) {
      for (const simple of compound.simples) {
        switch (simple.kind) {
          case 'id':
            a++;
            break;
          case 'class':
          case 'attribute':
            b++;
            break;
          case 'pseudo-class': {
            // :not() and :is() use the specificity of their argument
            // :where() has zero specificity
            const val = simple.value;
            if (val.startsWith('where(')) {
              // zero specificity
            } else if (val.startsWith('not(') || val.startsWith('is(')) {
              const inner = val.slice(val.indexOf('(') + 1, val.length - 1);
              const [ia, ib, ic] = computeSpecificity(inner);
              a += ia; b += ib; c += ic;
            } else {
              b++;
            }
            break;
          }
          case 'pseudo-element':
            c++;
            break;
          case 'type':
            c++;
            break;
          case 'universal':
            // zero specificity
            break;
        }
      }
    }

    return [a, b, c];

  } catch (error) {
    throw new Error(`[computeSpecificity] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Handles the gnosis compare Specificity workflow.
 */
export function compareSpecificity(a: Specificity, b: Specificity): number {
  try {
    if (a[0] !== b[0]) return a[0] - b[0];
    if (a[1] !== b[1]) return a[1] - b[1];
    return a[2] - b[2];

  } catch (error) {
    throw new Error(`[compareSpecificity] ${error instanceof Error ? error.message : String(error)}`);
  }
}

// ── Selector Matching ────────────────────────────────────────────────────────

interface NodeContext {
  node: CascadeNode;
  parent: NodeContext | null;
  /** Index among siblings (for :first-child etc.) */
  index: number;
  /** Previous sibling for + combinator */
  prevSibling: CascadeNode | null;
}

function getClasses(node: CascadeNode): string[] {
  try {
    const cn = node.props?.className ?? node.props?.class;
    if (typeof cn === 'string') return cn.split(/\s+/).filter(Boolean);
    return [];

  } catch (error) {
    throw new Error(`[getClasses] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function getId(node: CascadeNode): string | undefined {
  const id = node.props?.id;
  return typeof id === 'string' ? id : undefined;
}

function matchesCompound(compound: CompoundSelector, ctx: NodeContext): boolean {
  try {
    for (const simple of compound.simples) {
      if (!matchesSimple(simple, ctx)) return false;
    }
    return true;

  } catch (error) {
    throw new Error(`[matchesCompound] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function matchesSimple(simple: SimpleSelector, ctx: NodeContext): boolean {
  try {
    const { node } = ctx;
    switch (simple.kind) {
      case 'type':
        return node.type.toLowerCase() === simple.value.toLowerCase();
      case 'class':
        return getClasses(node).includes(simple.value);
      case 'id':
        return getId(node) === simple.value;
      case 'universal':
        return true;
      case 'attribute': {
        if (!simple.attr) return false;
        const attrVal = node.props?.[simple.attr];
        if (attrVal === undefined) return false;
        if (!simple.attrOp) return true; // [attr] -- just existence
        const v = String(attrVal);
        const expected = simple.attrValue ?? '';
        switch (simple.attrOp) {
          case '=': return v === expected;
          case '~=': return v.split(/\s+/).includes(expected);
          case '|=': return v === expected || v.startsWith(expected + '-');
          case '^=': return v.startsWith(expected);
          case '$=': return v.endsWith(expected);
          case '*=': return v.includes(expected);
          default: return false;
        }
      }
      case 'pseudo-class': {
        const val = simple.value;
        if (val === 'first-child') return ctx.index === 0;
        if (val === 'last-child') return true; // approximate -- would need sibling count
        if (val === 'hover' || val === 'focus' || val === 'active') return false;
        if (val === 'disabled') return node.props?.disabled === true;
        if (val === 'root') return ctx.parent === null;
        if (val.startsWith('not(')) {
          const inner = val.slice(4, val.length - 1);
          return !matchesSelector(inner, ctx);
        }
        if (val.startsWith('is(') || val.startsWith('where(')) {
          const inner = val.slice(val.indexOf('(') + 1, val.length - 1);
          // Split by comma and check if any match
          return inner.split(',').some(s => matchesSelector(s.trim(), ctx));
        }
        return false;
      }
      case 'pseudo-element':
        // Pseudo-elements don't match real nodes
        return false;
    }

  } catch (error) {
    throw new Error(`[matchesSimple] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function matchesSelector(selector: string, ctx: NodeContext): boolean {
  const seq = parseSelectorSequence(selector);
  return matchesSelectorSequence(seq, ctx);
}

function matchesSelectorSequence(seq: SelectorSequence, ctx: NodeContext): boolean {
  try {
    const { compounds, combinators } = seq;
    if (compounds.length === 0) return false;

    // Start from rightmost compound (the subject)
    const lastCompound = compounds[compounds.length - 1]!;
    if (!matchesCompound(lastCompound, ctx)) return false;

    // Walk leftward through combinators
    let currentCtx: NodeContext | null = ctx;
    for (let i = compounds.length - 2; i >= 0; i--) {
      const compound = compounds[i]!;
      const combinator = combinators[i]!;

      switch (combinator) {
        case 'descendant': {
          // Walk up ancestors until one matches
          let ancestor: NodeContext | null = currentCtx?.parent ?? null;
          let found = false;
          while (ancestor) {
            if (matchesCompound(compound, ancestor)) {
              currentCtx = ancestor;
              found = true;
              break;
            }
            ancestor = ancestor.parent;
          }
          if (!found) return false;
          break;
        }
        case 'child': {
          // Direct parent must match
          const parentNode: NodeContext | null = currentCtx?.parent ?? null;
          if (!parentNode || !matchesCompound(compound, parentNode)) return false;
          currentCtx = parentNode;
          break;
        }
        case 'adjacent': {
          // Previous sibling must match
          if (!currentCtx?.prevSibling) return false;
          const sibCtx: NodeContext = {
            node: currentCtx.prevSibling,
            parent: currentCtx.parent,
            index: currentCtx.index - 1,
            prevSibling: null,
          };
          if (!matchesCompound(compound, sibCtx)) return false;
          currentCtx = sibCtx;
          break;
        }
        case 'general-sibling': {
          // Any preceding sibling must match -- approximation: check prev only
          if (!currentCtx?.prevSibling) return false;
          const sibCtx: NodeContext = {
            node: currentCtx.prevSibling,
            parent: currentCtx.parent,
            index: currentCtx.index - 1,
            prevSibling: null,
          };
          if (!matchesCompound(compound, sibCtx)) return false;
          currentCtx = sibCtx;
          break;
        }
      }
    }

    return true;

  } catch (error) {
    throw new Error(`[matchesSelectorSequence] ${error instanceof Error ? error.message : String(error)}`);
  }
}

// ── Inherited Properties ─────────────────────────────────────────────────────

const INHERITED_PROPERTIES = new Set([
  'azimuth', 'border-collapse', 'border-spacing', 'caption-side', 'color',
  'cursor', 'direction', 'empty-cells', 'font', 'font-family', 'font-size',
  'font-style', 'font-variant', 'font-weight', 'letter-spacing', 'line-height',
  'list-style', 'list-style-image', 'list-style-position', 'list-style-type',
  'orphans', 'quotes', 'text-align', 'text-indent', 'text-transform',
  'visibility', 'white-space', 'widows', 'word-spacing', 'writing-mode',
  // SVG inherited properties
  'fill', 'fill-opacity', 'fill-rule', 'stroke', 'stroke-dasharray',
  'stroke-dashoffset', 'stroke-linecap', 'stroke-linejoin', 'stroke-miterlimit',
  'stroke-opacity', 'stroke-width', 'color-interpolation',
  'color-interpolation-filters', 'clip-rule', 'marker', 'marker-start',
  'marker-mid', 'marker-end', 'paint-order', 'shape-rendering',
  'text-rendering', 'pointer-events',
]);

/**
 * Returns whether is Inherited is true.
 */
export function isInherited(property: string): boolean {
  return INHERITED_PROPERTIES.has(property);
}

// ── Cascade Resolver ─────────────────────────────────────────────────────────

export interface CascadeResult {
  /** The tree with resolved styles attached to every node */
  tree: ResolvedNode;
  /** CSS rules that matched zero nodes -- dead CSS, safe to treeshake */
  deadRules: CSSRule[];
  /** CSS rules that matched at least one node */
  liveRules: CSSRule[];
  /** Total rules processed */
  totalRules: number;
  /** Nodes with style conflicts (multiple rules competing at equal specificity) */
  ambiguousNodes: { path: string; conflicts: PropertyConflict[] }[];
}

/**
 * Resolve the CSS cascade over a known tree.
 *
 * This is the core function. It takes a tree and CSS, walks every node,
 * matches every rule, sorts by specificity, resolves conflicts, propagates
 * inheritance, and returns the tree with resolved styles on every node.
 *
 * No ambiguity. No flying blind. The cascade is a function.
 */
export function resolveCascade(tree: CascadeNode, css: string | CSSRule[]): CascadeResult {
  const rules = typeof css === 'string' ? parseCSS(css) : css;
  const ruleMatchCounts = new Map<number, number>();
  const ambiguousNodes: CascadeResult['ambiguousNodes'] = [];

  // Initialize match counts
  for (const rule of rules) {
    ruleMatchCounts.set(rule.sourceOrder, 0);
  }

  function resolve(
    node: CascadeNode,
    parentCtx: NodeContext | null,
    parentStyles: Record<string, string>,
    siblingIndex: number,
    prevSibling: CascadeNode | null,
    path: string,
  ): ResolvedNode {
    try {
      const ctx: NodeContext = {
        node,
        parent: parentCtx,
        index: siblingIndex,
        prevSibling,
      };

      // FORK: find all rules that match this node
      const matched: MatchedRule[] = [];
      for (const rule of rules) {
        if (matchesSelector(rule.selector, ctx)) {
          ruleMatchCounts.set(rule.sourceOrder, (ruleMatchCounts.get(rule.sourceOrder) ?? 0) + 1);
          matched.push({
            selector: rule.selector,
            specificity: computeSpecificity(rule.selector),
            declarations: rule.declarations,
            sourceOrder: rule.sourceOrder,
          });
        }
      }

      // RACE + FOLD: for each property, find the winning declaration
      const propertyMap = new Map<string, {
        value: string;
        important: boolean;
        specificity: Specificity;
        sourceOrder: number;
        selector: string;
      }[]>();

      for (const match of matched) {
        for (const decl of match.declarations) {
          const existing = propertyMap.get(decl.property) ?? [];
          existing.push({
            value: decl.value,
            important: decl.important,
            specificity: match.specificity,
            sourceOrder: match.sourceOrder,
            selector: match.selector,
          });
          propertyMap.set(decl.property, existing);
        }
      }

      const resolvedStyles: Record<string, string> = {};
      const conflicts: PropertyConflict[] = [];

      for (const [property, candidates] of propertyMap) {
        // Sort: !important first, then specificity desc, then source order desc
        candidates.sort((a, b) => {
          if (a.important !== b.important) return a.important ? -1 : 1;
          const specCmp = compareSpecificity(b.specificity, a.specificity);
          if (specCmp !== 0) return specCmp;
          return b.sourceOrder - a.sourceOrder;
        });

        const winner = candidates[0]!;
        resolvedStyles[property] = winner.value;

        // Detect conflicts: multiple candidates with same specificity + importance
        if (candidates.length > 1) {
          const runner = candidates[1]!;
          const sameImportance = winner.important === runner.important;
          const sameSpecificity = compareSpecificity(winner.specificity, runner.specificity) === 0;
          if (sameImportance && sameSpecificity) {
            conflicts.push({
              property,
              contenders: candidates.map(c => ({
                selector: c.selector,
                specificity: c.specificity,
                value: c.value,
              })),
              winner: winner.value,
              reason: 'source-order',
            });
          }
        }
      }

      // TRACE: inheritance -- for inherited properties not explicitly set,
      // pull from parent's resolved styles
      for (const [prop, val] of Object.entries(parentStyles)) {
        if (isInherited(prop) && !(prop in resolvedStyles)) {
          resolvedStyles[prop] = val;
        }
      }

      // Also apply inline styles from props.style (highest priority, like specificity ∞)
      const inlineStyle = node.props?.style;
      if (inlineStyle && typeof inlineStyle === 'object') {
        for (const [prop, val] of Object.entries(inlineStyle as Record<string, unknown>)) {
          if (val !== undefined && val !== null) {
            // Convert camelCase to kebab-case
            const kebab = prop.replace(/([A-Z])/g, '-$1').toLowerCase();
            resolvedStyles[kebab] = String(val);
          }
        }
      }

      // Also apply HTML presentational attributes (SVG: fill, stroke, etc.)
      const presentationalAttrs = [
        'fill', 'stroke', 'stroke-width', 'stroke-linecap', 'stroke-linejoin',
        'stroke-dasharray', 'stroke-dashoffset', 'stroke-miterlimit',
        'stroke-opacity', 'fill-opacity', 'fill-rule', 'clip-rule',
        'opacity', 'transform', 'color', 'font-size', 'font-family',
        'text-anchor', 'dominant-baseline', 'width', 'height',
      ];
      // Map camelCase React prop names to kebab-case
      const camelToKebab: Record<string, string> = {
        strokeWidth: 'stroke-width',
        strokeLinecap: 'stroke-linecap',
        strokeLinejoin: 'stroke-linejoin',
        strokeDasharray: 'stroke-dasharray',
        strokeDashoffset: 'stroke-dashoffset',
        strokeMiterlimit: 'stroke-miterlimit',
        strokeOpacity: 'stroke-opacity',
        fillOpacity: 'fill-opacity',
        fillRule: 'fill-rule',
        clipRule: 'clip-rule',
        fontSize: 'font-size',
        fontFamily: 'font-family',
        textAnchor: 'text-anchor',
        dominantBaseline: 'dominant-baseline',
      };

      if (node.props) {
        for (const [prop, val] of Object.entries(node.props)) {
          const kebab = camelToKebab[prop] ?? prop;
          if (presentationalAttrs.includes(kebab) && val !== undefined && val !== null) {
            // Presentational attributes have lower priority than CSS
            // Only apply if not already set by CSS
            if (!(kebab in resolvedStyles)) {
              resolvedStyles[kebab] = String(val);
            }
          }
        }
      }

      if (conflicts.length > 0) {
        ambiguousNodes.push({ path, conflicts });
      }

      // Recurse into children
      const resolvedChildren: (ResolvedNode | string)[] = [];
      if (node.children) {
        let prevChild: CascadeNode | null = null;
        for (let i = 0; i < node.children.length; i++) {
          const child = node.children[i]!;
          if (typeof child === 'string') {
            resolvedChildren.push(child);
          } else {
            resolvedChildren.push(
              resolve(child, ctx, resolvedStyles, i, prevChild, `${path} > ${child.type}`)
            );
            prevChild = child;
          }
        }
      }

      return {
        type: node.type,
        props: node.props,
        children: resolvedChildren.length > 0 ? resolvedChildren : undefined,
        resolvedStyles,
        matchedRules: matched,
        conflicts,
      };
  
    } catch (error) {
      throw new Error(`[resolve] ${error instanceof Error ? error.message : String(error)}`);
    }
}

  const resolved = resolve(tree, null, {}, 0, null, tree.type);

  // Partition rules into live and dead
  const deadRules: CSSRule[] = [];
  const liveRules: CSSRule[] = [];
  for (const rule of rules) {
    if ((ruleMatchCounts.get(rule.sourceOrder) ?? 0) > 0) {
      liveRules.push(rule);
    } else {
      deadRules.push(rule);
    }
  }

  return {
    tree: resolved,
    deadRules,
    liveRules,
    totalRules: rules.length,
    ambiguousNodes,
  };
}

// ── Treeshaker ───────────────────────────────────────────────────────────────

/**
 * Given CSS and a tree, return only the CSS rules that actually match
 * at least one node. Dead rules are eliminated.
 */
export function treeshakeCSS(tree: CascadeNode, css: string): string {
  const rules = parseCSS(css);
  const { liveRules } = resolveCascade(tree, rules);

  // Reconstruct CSS from live rules only
  return liveRules
    .sort((a, b) => a.sourceOrder - b.sourceOrder)
    .map(rule => {
      const decls = rule.declarations
        .map(d => `  ${d.property}: ${d.value}${d.important ? ' !important' : ''};`)
        .join('\n');
      return `${rule.selector} {\n${decls}\n}`;
    })
    .join('\n\n');
}

// ── Diagnostic Formatter ─────────────────────────────────────────────────────

export interface CascadeDiagnostic {
  path: string;
  property: string;
  winner: { selector: string; specificity: string; value: string };
  losers: { selector: string; specificity: string; value: string }[];
  reason: 'specificity' | 'source-order' | 'important';
}

/**
 * Explain the cascade at a specific node path.
 * Returns a human-readable breakdown of every property, who set it, and why.
 */
export function explainNode(result: CascadeResult, path: string[]): string {
  try {
    const node = walkPath(result.tree, path);
    if (!node) return `Node not found at path: ${path.join(' > ')}`;

    const lines: string[] = [`Resolved styles for <${node.type}>:`];
    lines.push('');

    // Group by source
    const fromCSS: string[] = [];
    const fromInherit: string[] = [];
    const fromAttr: string[] = [];

    for (const [prop, val] of Object.entries(node.resolvedStyles)) {
      // Find the winning rule for this property (highest specificity, then source order)
      const candidates = node.matchedRules
        .filter(r => r.declarations.some(d => d.property === prop))
        .sort((a, b) => {
          const specCmp = compareSpecificity(b.specificity, a.specificity);
          if (specCmp !== 0) return specCmp;
          return b.sourceOrder - a.sourceOrder;
        });
      const matchingRule = candidates[0];
      if (matchingRule) {
        fromCSS.push(`  ${prop}: ${val}  ← ${matchingRule.selector} [${matchingRule.specificity.join(',')}]`);
      } else {
        fromInherit.push(`  ${prop}: ${val}  ← inherited`);
      }
    }

    if (fromCSS.length > 0) {
      lines.push('From CSS rules:');
      lines.push(...fromCSS);
    }
    if (fromInherit.length > 0) {
      lines.push('');
      lines.push('Inherited from parent:');
      lines.push(...fromInherit);
    }
    if (fromAttr.length > 0) {
      lines.push('');
      lines.push('From HTML/SVG attributes:');
      lines.push(...fromAttr);
    }

    if (node.conflicts.length > 0) {
      lines.push('');
      lines.push('CASCADE CONFLICTS:');
      for (const c of node.conflicts) {
        lines.push(`  ${c.property}: "${c.winner}" won by ${c.reason}`);
        for (const contender of c.contenders) {
          const marker = contender.value === c.winner ? '  [v]' : '  [x]';
          lines.push(`  ${marker} ${contender.selector} [${contender.specificity.join(',')}] → "${contender.value}"`);
        }
      }
    }

    return lines.join('\n');

  } catch (error) {
    throw new Error(`[explainNode] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function walkPath(node: ResolvedNode, path: string[]): ResolvedNode | null {
  try {
    if (path.length === 0) return node;
    const [next, ...rest] = path;
    if (!node.children) return null;
    for (const child of node.children) {
      if (typeof child === 'string') continue;
      if (child.type === next) {
        return walkPath(child, rest);
      }
    }
    return null;

  } catch (error) {
    throw new Error(`[walkPath] ${error instanceof Error ? error.message : String(error)}`);
  }
}

// ── Topology View ────────────────────────────────────────────────────────────

/**
 * Express the cascade as fork/race/fold morphisms.
 * Returns a string diagram description of the cascade topology.
 */
export function cascadeTopology(result: CascadeResult): string {
  try {
    const lines: string[] = ['CSS Cascade Topology:', ''];

    lines.push(`  ${result.totalRules} rules ──FORK──→ selector matching`);
    lines.push(`  ${result.liveRules.length} live rules ──RACE──→ specificity competition`);
    lines.push(`  ${result.deadRules.length} dead rules ──VENT──→ treeshaken`);
    lines.push('');

    if (result.ambiguousNodes.length > 0) {
      lines.push('  Ambiguous nodes (source-order decided):');
      for (const { path, conflicts } of result.ambiguousNodes) {
        for (const c of conflicts) {
          lines.push(`    ${path} → ${c.property}: ${c.contenders.length} rules tied at same specificity`);
        }
      }
    } else {
      lines.push('  No ambiguity: every property resolved by specificity alone.');
    }

    if (result.deadRules.length > 0) {
      lines.push('');
      lines.push('  Dead CSS (matched zero nodes):');
      for (const rule of result.deadRules) {
        lines.push(`    ${rule.selector}`);
      }
    }

    return lines.join('\n');

  } catch (error) {
    throw new Error(`[cascadeTopology] ${error instanceof Error ? error.message : String(error)}`);
  }
}
