// @a0n/.ignored_gnosis/src/css-layout-analysis.ts
/**
 * CSS Layout Thermodynamics Analysis
 *
 * Runs the seven CSS layout detectors from the polyglot scanner on
 * TSX/JSX source text and converts findings into SemanticObligations
 * and Diagnostics for the gnosis cross-domain pipeline.
 *
 * Each detector is backed by a Lean 4 theorem from
 * CSSLayoutThermodynamics.lean. The theorems prove the invariants;
 * these detectors find violations.
 *
 * @module @a0n/gnosis/css-layout-analysis
 */

import type { Diagnostic, DiagnosticCode } from './betty/compiler';
import type { SemanticObligation } from './betty/semantic-compatibility';
import {
  compileStructuredGraph,
  extractStructuredDeclarations,
} from './structured-data';

// ── Types ────────────────────────────────────────────────────────────

export interface CSSLayoutFix {
  readonly line: number;
  readonly oldText: string;
  readonly newText: string;
  readonly description: string;
}

export interface CSSLayoutFinding {
  readonly rule: string;
  readonly line: number;
  readonly message: string;
  readonly level: 'error' | 'warning' | 'hint';
  readonly axis?: string;
  readonly crossingNumber?: number;
  readonly beta1?: number;
  readonly fix?: CSSLayoutFix;
}

export interface CSSLayoutAnalysisResult {
  readonly findings: readonly CSSLayoutFinding[];
  readonly diagnostics: readonly Diagnostic[];
  readonly obligations: readonly SemanticObligation[];
}

/**
 * Apply all autofixes to source text. Returns the fixed source.
 * Applies fixes in reverse line order to preserve line numbers.
 */
export function applyFixes(
  sourceText: string,
  findings: readonly CSSLayoutFinding[]
): string {
  try {
    const fixes = findings
      .filter((f): f is CSSLayoutFinding & { fix: CSSLayoutFix } => !!f.fix)
      .sort((a, b) => b.fix.line - a.fix.line); // reverse order

    const lines = sourceText.split('\n');
    for (const { fix } of fixes) {
      const idx = fix.line - 1;
      if (idx >= 0 && idx < lines.length && lines[idx]!.includes(fix.oldText)) {
        lines[idx] = lines[idx]!.replace(fix.oldText, fix.newText);
      }
    }
    return lines.join('\n');
  } catch (error) {
    throw new Error(
      `[applyFixes] ${error instanceof Error ? error.message : String(error)}`
    );
  }
}

// ── Tailwind class parsing ───────────────────────────────────────────

const TAILWIND_AXIS_MAP: Record<string, string> = {
  'w-full': 'width',
  'w-screen': 'width',
  'w-auto': 'width',
  'w-fit': 'width',
  'w-min': 'width',
  'w-max': 'width',
  'w-dvw': 'width',
  'w-svw': 'width',
  'h-full': 'height',
  'h-screen': 'height',
  'h-auto': 'height',
  'h-fit': 'height',
  'h-min': 'height',
  'h-max': 'height',
  'h-dvh': 'height',
  'h-svh': 'height',
  block: 'display',
  inline: 'display',
  'inline-block': 'display',
  flex: 'display',
  'inline-flex': 'display',
  grid: 'display',
  'inline-grid': 'display',
  hidden: 'display',
  contents: 'display',
  table: 'display',
  'flow-root': 'display',
  static: 'position',
  relative: 'position',
  absolute: 'position',
  fixed: 'position',
  sticky: 'position',
  'overflow-auto': 'overflow',
  'overflow-hidden': 'overflow',
  'overflow-visible': 'overflow',
  'overflow-scroll': 'overflow',
  'overflow-clip': 'overflow',
  'flex-row': 'flex-direction',
  'flex-col': 'flex-direction',
  'flex-row-reverse': 'flex-direction',
  'flex-col-reverse': 'flex-direction',
  'flex-wrap': 'flex-wrap',
  'flex-nowrap': 'flex-wrap',
  'flex-wrap-reverse': 'flex-wrap',
  'justify-start': 'justify-content',
  'justify-end': 'justify-content',
  'justify-center': 'justify-content',
  'justify-between': 'justify-content',
  'justify-around': 'justify-content',
  'justify-evenly': 'justify-content',
  'items-start': 'align-items',
  'items-end': 'align-items',
  'items-center': 'align-items',
  'items-baseline': 'align-items',
  'items-stretch': 'align-items',
  'text-left': 'text-align',
  'text-center': 'text-align',
  'text-right': 'text-align',
  'text-justify': 'text-align',
};

const NUMERIC_AXES: Array<{ prefix: string; axis: string }> = [
  { prefix: 'w-', axis: 'width' },
  { prefix: 'h-', axis: 'height' },
  { prefix: 'min-w-', axis: 'min-width' },
  { prefix: 'min-h-', axis: 'min-height' },
  { prefix: 'max-w-', axis: 'max-width' },
  { prefix: 'max-h-', axis: 'max-height' },
  { prefix: 'p-', axis: 'padding' },
  { prefix: 'px-', axis: 'padding-x' },
  { prefix: 'py-', axis: 'padding-y' },
  { prefix: 'pt-', axis: 'padding-top' },
  { prefix: 'pb-', axis: 'padding-bottom' },
  { prefix: 'pl-', axis: 'padding-left' },
  { prefix: 'pr-', axis: 'padding-right' },
  { prefix: 'm-', axis: 'margin' },
  { prefix: 'mx-', axis: 'margin-x' },
  { prefix: 'my-', axis: 'margin-y' },
  { prefix: 'mt-', axis: 'margin-top' },
  { prefix: 'mb-', axis: 'margin-bottom' },
  { prefix: 'ml-', axis: 'margin-left' },
  { prefix: 'mr-', axis: 'margin-right' },
  // gap-x and gap-y are INDEPENDENT axes (column-gap vs row-gap), not conflicts
  { prefix: 'gap-x-', axis: 'column-gap' },
  { prefix: 'gap-y-', axis: 'row-gap' },
  { prefix: 'gap-', axis: 'gap' },
  { prefix: 'z-', axis: 'z-index' },
  // Per-corner rounded overrides (rounded-tl-*, rounded-br-*) are independent of rounded-*
  // Only flag shorthand rounded-* vs rounded-* conflicts, not corner-specific overrides
];

const TEXT_SIZE_SUFFIXES = new Set([
  'xs',
  'sm',
  'base',
  'lg',
  'xl',
  '2xl',
  '3xl',
  '4xl',
  '5xl',
  '6xl',
  '7xl',
  '8xl',
  '9xl',
]);

const BREAKPOINT_ORDER = ['sm', 'md', 'lg', 'xl', '2xl'];

function classifyAxis(cls: string): string | null {
  try {
    const bare = cls.replace(
      /^(?:sm|md|lg|xl|2xl|hover|focus|active|group-hover|disabled|first|last|odd|even|placeholder|checked|required|invalid|peer-\w+|dark|print|motion-safe|motion-reduce|contrast-more|contrast-less|forced-colors|focus-within|focus-visible):/,
      ''
    );
    const unsigned = bare.replace(/^-/, '');

    // Per-corner border-radius is independent (rounded-tl-sm doesn't conflict with rounded-lg)
    if (/^rounded-(?:tl|tr|bl|br|t|b|l|r|s|e|ss|se|es|ee)-/.test(unsigned))
      return `border-radius-${unsigned.split('-')[1]}`;
    if (/^rounded-/.test(unsigned)) return 'border-radius';
    if (TAILWIND_AXIS_MAP[unsigned]) return TAILWIND_AXIS_MAP[unsigned]!;
    if (unsigned.startsWith('text-')) {
      const suffix = unsigned.slice(5);
      if (TEXT_SIZE_SUFFIXES.has(suffix)) return 'font-size';
      if (/^\[/.test(suffix)) {
        if (
          /^\[(?:var\(|#|rgb|hsl|oklch|oklab|lch|lab|hwb|color\()/.test(suffix)
        )
          return 'text-color';
        return 'font-size';
      }
      return 'text-color';
    }
    for (const { prefix, axis } of NUMERIC_AXES) {
      if (unsigned.startsWith(prefix) && unsigned.length > prefix.length)
        return axis;
    }
    return null;
  } catch (error) {
    throw new Error(
      `[classifyAxis] ${error instanceof Error ? error.message : String(error)}`
    );
  }
}

function extractResponsivePrefix(cls: string): string | null {
  const match = cls.match(/^(sm|md|lg|xl|2xl):/);
  return match ? match[1]! : null;
}

function extractVariantKey(cls: string): string {
  try {
    const prefixes: string[] = [];
    let rest = cls;
    while (true) {
      const match = rest.match(
        /^(sm|md|lg|xl|2xl|hover|focus|active|group-hover|dark|disabled|first|last|odd|even|placeholder|checked|required|invalid|focus-within|focus-visible|peer-\w+|motion-safe|motion-reduce|contrast-more|contrast-less|forced-colors|print):/
      );
      if (!match) break;
      prefixes.push(match[1]!);
      rest = rest.slice(match[0].length);
    }
    return prefixes.sort().join(':') || '';
  } catch (error) {
    throw new Error(
      `[extractVariantKey] ${
        error instanceof Error ? error.message : String(error)
      }`
    );
  }
}

function extractClasses(line: string): string[] {
  const classes: string[] = [];
  const patterns = [
    /(?:className|class|tw)\s*=\s*"([^"]*)"/g,
    /(?:className|class|tw)\s*=\s*'([^']*)'/g,
    /(?:className|class|tw)\s*=\s*\{`([^`]*)`\}/g,
    /cn\s*\(\s*["'`]([^"'`]*)["'`]/g,
    /clsx\s*\(\s*["'`]([^"'`]*)["'`]/g,
    /twMerge\s*\(\s*["'`]([^"'`]*)["'`]/g,
  ];
  for (const pattern of patterns) {
    let match;
    while ((match = pattern.exec(line)) !== null) {
      for (const cls of match[1]!.split(/\s+/).filter(Boolean))
        classes.push(cls);
    }
  }
  return classes;
}

// ── Finding-to-diagnostic mapping ────────────────────────────────────

const RULE_TO_CODE: Record<string, DiagnosticCode> = {
  TAILWIND_CONFLICT: 'LAYOUT_TAILWIND_CONFLICT',
  FLEX_OVERFLOW: 'LAYOUT_FLEX_OVERFLOW',
  POSITION_ESCAPE: 'LAYOUT_POSITION_ESCAPE',
  RESPONSIVE_NON_MONOTONE: 'LAYOUT_RESPONSIVE_NON_MONOTONE',
  MARGIN_COLLAPSE_RISK: 'LAYOUT_MARGIN_COLLAPSE',
  MISSING_FLEX_PARENT: 'LAYOUT_MISSING_FLEX_PARENT',
  OVERFLOW_UNHANDLED: 'LAYOUT_OVERFLOW_UNHANDLED',
  SEO_HEADING_HIERARCHY: 'SEO_HEADING_HIERARCHY',
  SEO_MISSING_META: 'SEO_MISSING_META',
  SEO_MISSING_STRUCTURED_DATA: 'SEO_MISSING_STRUCTURED_DATA',
  A11Y_MISSING_ALT: 'A11Y_MISSING_ALT',
  A11Y_MISSING_LABEL: 'A11Y_MISSING_LABEL',
  A11Y_COLOR_CONTRAST: 'A11Y_COLOR_CONTRAST',
  BOUNDARY_SERVER_IMPORT_IN_CLIENT: 'BOUNDARY_SERVER_IMPORT_IN_CLIENT',
  BOUNDARY_SECRET_IN_CLIENT: 'BOUNDARY_SECRET_IN_CLIENT',
  CACHE_IMMUTABLE_MUTABLE_RESOURCE: 'CACHE_IMMUTABLE_MUTABLE_RESOURCE',
  SLOP_EMOJI_IN_CODE: 'SLOP_EMOJI_IN_CODE',
  SLOP_CONSOLE_LOG_IN_PROD: 'SLOP_CONSOLE_LOG_IN_PROD',
  SLOP_TODO_IN_PROD: 'SLOP_TODO_IN_PROD',
  SLOP_HARDCODED_URL: 'SLOP_HARDCODED_URL',
};

const LEVEL_TO_SEVERITY: Record<string, 'error' | 'warning' | 'info'> = {
  error: 'error',
  warning: 'warning',
  hint: 'info',
};

function findingToDiagnostic(f: CSSLayoutFinding): Diagnostic {
  const diagnostic: Diagnostic = {
    line: f.line,
    column: 0,
    message: f.message,
    severity: LEVEL_TO_SEVERITY[f.level] ?? 'warning',
  };
  const code = RULE_TO_CODE[f.rule];
  if (code !== undefined) {
    diagnostic.code = code;
  }
  return diagnostic;
}

function findingToObligation(f: CSSLayoutFinding): SemanticObligation {
  return {
    key: `layout-thermodynamics:${f.rule}`,
    message: f.message,
    severity: f.level === 'hint' ? 'info' : f.level,
    facetKey: f.rule,
  };
}

// ── Detectors ────────────────────────────────────────────────────────

// Classes that are NOT in the text-color axis despite starting with text-
const TEXT_NON_COLOR = new Set([
  'text-ellipsis',
  'text-clip',
  'text-nowrap',
  'text-wrap',
  'text-balance',
  'text-pretty',
  'text-start',
  'text-end',
]);

// Intentional multi-class patterns that are not conflicts
function isIntentionalPattern(
  a: string,
  b: string,
  classes: string[]
): boolean {
  try {
    // bg-clip-text + text-transparent is a gradient text pattern
    if (
      (a === 'text-transparent' || b === 'text-transparent') &&
      classes.includes('bg-clip-text')
    )
      return true;
    return false;
  } catch (error) {
    throw new Error(
      `[isIntentionalPattern] ${
        error instanceof Error ? error.message : String(error)
      }`
    );
  }
}

function detectTailwindConflicts(lines: string[]): CSSLayoutFinding[] {
  try {
    const findings: CSSLayoutFinding[] = [];
    for (let i = 0; i < lines.length; i++) {
      const line = lines[i]!;
      const classes = extractClasses(line);
      if (classes.length < 2) continue;
      const axisSeen = new Map<string, string>();
      for (const cls of classes) {
        if (
          TEXT_NON_COLOR.has(
            cls.replace(/^(?:sm|md|lg|xl|2xl|dark|hover|focus):/, '')
          )
        )
          continue;
        const variantKey = extractVariantKey(cls);
        const axis = classifyAxis(cls);
        if (!axis) continue;
        const key = `${variantKey}:${axis}`;
        const existing = axisSeen.get(key);
        if (existing && existing !== cls) {
          if (isIntentionalPattern(existing, cls, classes)) continue;
          // Autofix: remove the first (redundant) class, keep the later one (Tailwind last-wins)
          const fixOld = existing;
          const fixNew = '';
          findings.push({
            rule: 'TAILWIND_CONFLICT',
            line: i + 1,
            message: `Conflicting classes '${existing}' and '${cls}' both target '${axis}'${
              variantKey ? ` in variant '${variantKey}'` : ''
            }. Per tailwind_no_self_conflict.`,
            level: 'warning',
            axis,
            crossingNumber: 2,
            beta1: 0,
            fix: {
              line: i + 1,
              oldText: `${fixOld} `,
              newText: fixNew,
              description: `Remove redundant '${fixOld}' (conflicts with '${cls}')`,
            },
          });
        }
        axisSeen.set(key, cls);
      }
    }
    return findings;
  } catch (error) {
    throw new Error(
      `[detectTailwindConflicts] ${
        error instanceof Error ? error.message : String(error)
      }`
    );
  }
}

function detectFlexOverflow(lines: string[]): CSSLayoutFinding[] {
  try {
    const findings: CSSLayoutFinding[] = [];
    for (let i = 0; i < lines.length; i++) {
      const classes = new Set(extractClasses(lines[i]!));
      if (!classes.has('flex') && !classes.has('inline-flex')) continue;
      if (classes.has('flex-wrap') || classes.has('flex-wrap-reverse'))
        continue;
      if (
        classes.has('overflow-hidden') ||
        classes.has('overflow-auto') ||
        classes.has('overflow-scroll') ||
        classes.has('overflow-clip')
      )
        continue;
      const next = lines.slice(i + 1, Math.min(i + 20, lines.length)).join(' ');
      // Only count substantial fixed widths (w-20+, ~80px+), not icon sizes (w-4, w-5, w-8)
      const fixedMatches = (next.match(/\bw-(\d+)\b/g) || []) as string[];
      const fixed = fixedMatches.filter(
        (m) => parseInt(m.replace('w-', ''), 10) >= 20
      ).length;
      const hasFlexAbsorber =
        /\bflex-1\b|\bflex-auto\b|\bmin-w-0\b|\btruncate\b/.test(next);
      if (fixed >= 3 && !hasFlexAbsorber) {
        // Autofix: add flex-wrap after flex
        const flexMatch = lines[i]!.match(/\bflex\b/);
        const fixOld = flexMatch ? 'flex' : 'inline-flex';
        findings.push({
          rule: 'FLEX_OVERFLOW',
          line: i + 1,
          message: `Flex container with ${fixed} fixed-width children, no wrap/overflow. Per flex_overflow_condition.`,
          level: 'warning',
          crossingNumber: fixed,
          fix: {
            line: i + 1,
            oldText: fixOld,
            newText: `${fixOld} flex-wrap`,
            description: 'Add flex-wrap to prevent overflow',
          },
        });
      }
    }
    return findings;
  } catch (error) {
    throw new Error(
      `[detectFlexOverflow] ${
        error instanceof Error ? error.message : String(error)
      }`
    );
  }
}

function detectPositionEscape(lines: string[]): CSSLayoutFinding[] {
  try {
    const findings: CSSLayoutFinding[] = [];
    for (let i = 0; i < lines.length; i++) {
      const classes = extractClasses(lines[i]!);
      if (!classes.includes('absolute')) continue;
      // fixed/sticky on same element is self-positioning, not escape
      if (classes.includes('fixed') || classes.includes('sticky')) continue;
      let found = false;
      // Scan upward for a positioned ancestor. Prior heuristic used JSX depth while walking
      // backward, which false-positived nested component trees (e.g. Motion wrappers).
      // TSX files often exceed 120 lines between a root `relative` and nested `absolute` overlays.
      for (let j = i - 1; j >= Math.max(0, i - 250); j--) {
        const line = lines[j]!;
        const prev = extractClasses(line);
        if (
          prev.some((c) => c === 'relative' || c === 'fixed' || c === 'sticky')
        ) {
          found = true;
          break;
        }
        // Also check raw line for 'relative' in template literals the extractor can't parse
        if (/\brelative\b/.test(line)) {
          found = true;
          break;
        }
      }
      if (!found) {
        // Suppress for component root returns (single-element components designed for composition)
        const isReturnStatement =
          i > 0 && /^\s*return\s*\(?\s*$/.test(lines[i - 1]?.trim() ?? '');
        if (isReturnStatement) continue;
        findings.push({
          rule: 'POSITION_ESCAPE',
          line: i + 1,
          message: `'absolute' without positioned ancestor. Per all_static_means_escape: escapes to viewport.`,
          level: 'warning',
          crossingNumber: 1,
        });
      }
    }
    return findings;
  } catch (error) {
    throw new Error(
      `[detectPositionEscape] ${
        error instanceof Error ? error.message : String(error)
      }`
    );
  }
}

function detectResponsiveNonMonotone(lines: string[]): CSSLayoutFinding[] {
  try {
    const findings: CSSLayoutFinding[] = [];
    for (let i = 0; i < lines.length; i++) {
      const classes = extractClasses(lines[i]!);
      if (classes.length === 0) continue;
      const vis = new Map<string, boolean>();
      let hasResp = false;
      for (const cls of classes) {
        const bp = extractResponsivePrefix(cls);
        const bare = cls.replace(/^(?:sm|md|lg|xl|2xl):/, '');
        if (
          [
            'hidden',
            'block',
            'flex',
            'grid',
            'inline',
            'inline-block',
            'inline-flex',
          ].includes(bare)
        ) {
          vis.set(bp ?? 'base', bare !== 'hidden');
          if (bp) hasResp = true;
        }
      }
      if (!hasResp || vis.size < 3) continue;
      const chain: boolean[] = [];
      const base = vis.get('base');
      if (base !== undefined) chain.push(base);
      for (const bp of BREAKPOINT_ORDER) {
        const v = vis.get(bp);
        if (v !== undefined) chain.push(v);
      }
      let transitions = 0;
      for (let k = 1; k < chain.length; k++) {
        if (chain[k] !== chain[k - 1]) transitions++;
      }
      if (transitions >= 2) {
        findings.push({
          rule: 'RESPONSIVE_NON_MONOTONE',
          line: i + 1,
          message: `Non-monotone visibility (${transitions} transitions). Per isMonotoneVisibility.`,
          level: 'warning',
          crossingNumber: transitions,
          beta1: 1,
        });
      }
    }
    return findings;
  } catch (error) {
    throw new Error(
      `[detectResponsiveNonMonotone] ${
        error instanceof Error ? error.message : String(error)
      }`
    );
  }
}

function detectMarginCollapse(lines: string[]): CSSLayoutFinding[] {
  try {
    const findings: CSSLayoutFinding[] = [];
    for (let i = 1; i < lines.length; i++) {
      const prevC = extractClasses(lines[i - 1]!);
      const currC = extractClasses(lines[i]!);
      const mb = prevC.find(
        (c) => /^mb-(\d+)/.test(c) && parseInt(c.replace('mb-', ''), 10) >= 8
      );
      const mt = currC.find(
        (c) => /^mt-(\d+)/.test(c) && parseInt(c.replace('mt-', ''), 10) >= 8
      );
      if (mb && mt) {
        findings.push({
          rule: 'MARGIN_COLLAPSE_RISK',
          line: i + 1,
          message: `Adjacent '${mb}' and '${mt}' may collapse. Per margin_collapse_monotone: max(m1,m2) not m1+m2.`,
          level: 'hint',
        });
      }
    }
    return findings;
  } catch (error) {
    throw new Error(
      `[detectMarginCollapse] ${
        error instanceof Error ? error.message : String(error)
      }`
    );
  }
}

function detectMissingFlexParent(lines: string[]): CSSLayoutFinding[] {
  try {
    const findings: CSSLayoutFinding[] = [];
    const FLEX_CHILD = new Set([
      'flex-1',
      'flex-auto',
      'flex-initial',
      'flex-none',
      'grow',
      'grow-0',
      'shrink',
      'shrink-0',
      'basis-0',
      'basis-full',
      'self-start',
      'self-end',
      'self-center',
      'self-stretch',
      'self-auto',
    ]);
    for (let i = 0; i < lines.length; i++) {
      const classes = extractClasses(lines[i]!);
      if (classes.length === 0) continue;
      const child = classes.find(
        (c) => FLEX_CHILD.has(c) || /^basis-\d/.test(c)
      );
      if (!child) continue;
      if (
        classes.some((c) =>
          ['flex', 'inline-flex', 'grid', 'inline-grid'].includes(c)
        )
      )
        continue;
      let found = false;
      for (let j = i - 1; j >= Math.max(0, i - 50); j--) {
        const prevClasses = extractClasses(lines[j]!);
        if (
          prevClasses.some((c) =>
            ['flex', 'inline-flex', 'grid', 'inline-grid'].includes(c)
          )
        ) {
          found = true;
          break;
        }
        // Also check raw line for flex in template literals or CSS classes
        if (/\bflex\b|\bgrid\b/.test(lines[j]!)) {
          found = true;
          break;
        }
        if (lines[j]!.includes('</') && !lines[j]!.includes('<')) break;
      }
      if (!found) {
        findings.push({
          rule: 'MISSING_FLEX_PARENT',
          line: i + 1,
          message: `'${child}' without flex/grid parent within 15 lines. Per flex_child_at_least_basis.`,
          level: 'hint',
        });
      }
    }
    return findings;
  } catch (error) {
    throw new Error(
      `[detectMissingFlexParent] ${
        error instanceof Error ? error.message : String(error)
      }`
    );
  }
}

function detectOverflowUnhandled(lines: string[]): CSSLayoutFinding[] {
  try {
    const findings: CSSLayoutFinding[] = [];
    for (let i = 0; i < lines.length; i++) {
      const classes = extractClasses(lines[i]!);
      if (classes.length === 0) continue;
      const classSet = new Set(classes);
      // Only flag meaningful container heights (h-12+, h-screen, etc.), not icon sizes (h-3..h-10)
      const hasFixedH = classes.some((c) => {
        if (['h-screen', 'h-dvh', 'h-svh', 'h-full'].includes(c)) return true;
        const m = c.match(/^h-(\d+)/);
        return m ? parseInt(m[1]!, 10) >= 12 : false;
      });
      const hasOverflow =
        classSet.has('overflow-auto') ||
        classSet.has('overflow-hidden') ||
        classSet.has('overflow-scroll') ||
        classSet.has('overflow-y-auto') ||
        classSet.has('overflow-y-hidden') ||
        classSet.has('overflow-y-scroll') ||
        classSet.has('overflow-clip');
      if (hasFixedH && !hasOverflow) {
        const next = lines
          .slice(i + 1, Math.min(i + 10, lines.length))
          .join(' ');
        if (/\.map\s*\(|\{children\}|\{props\.children\}/.test(next)) {
          const hCls = classes.find((c) => /^h-/.test(c)) ?? 'h-*';
          findings.push({
            rule: 'OVERFLOW_UNHANDLED',
            line: i + 1,
            message: `Fixed-height '${hCls}' with dynamic content, no overflow handling. Per css_box_width_nonneg.`,
            level: 'warning',
            crossingNumber: 1,
            fix: {
              line: i + 1,
              oldText: hCls,
              newText: `${hCls} overflow-auto`,
              description: `Add overflow-auto to '${hCls}' container`,
            },
          });
        }
      }
    }
    return findings;
  } catch (error) {
    throw new Error(
      `[detectOverflowUnhandled] ${
        error instanceof Error ? error.message : String(error)
      }`
    );
  }
}

// ═══════════════════════════════════════════════════════════════════════
// SEO / Structured Data / Meta Detectors
// ═══════════════════════════════════════════════════════════════════════

const HEADING_LEVELS: Record<string, number> = {
  h1: 1,
  h2: 2,
  h3: 3,
  h4: 4,
  h5: 5,
  h6: 6,
};

/** SEO_HEADING_HIERARCHY: headings must descend (h1 > h2 > h3), never skip or reverse */
function detectHeadingHierarchy(lines: string[]): CSSLayoutFinding[] {
  try {
    const findings: CSSLayoutFinding[] = [];
    let lastLevel = 0;
    let h1Count = 0;
    for (let i = 0; i < lines.length; i++) {
      const line = lines[i]!;
      const match = line.match(/<(h[1-6])\b/);
      if (!match) continue;
      const tag = match[1]!;
      const level = HEADING_LEVELS[tag]!;
      if (level === 1) h1Count++;
      if (h1Count > 1) {
        findings.push({
          rule: 'SEO_HEADING_HIERARCHY',
          line: i + 1,
          message: `Multiple <h1> tags (${h1Count}). Only one <h1> per page for SEO. Use <h2>+ for subsections.`,
          level: 'warning',
          crossingNumber: h1Count,
        });
      }
      if (lastLevel > 0 && level > lastLevel + 1) {
        findings.push({
          rule: 'SEO_HEADING_HIERARCHY',
          line: i + 1,
          message: `Heading <${tag}> skips level (last was <h${lastLevel}>). Headings must descend: h${lastLevel} -> h${
            lastLevel + 1
          }.`,
          level: 'warning',
          crossingNumber: level - lastLevel,
        });
      }
      lastLevel = level;
    }
    return findings;
  } catch (error) {
    throw new Error(
      `[detectHeadingHierarchy] ${
        error instanceof Error ? error.message : String(error)
      }`
    );
  }
}

/** SEO_MISSING_META: pages should have title, description, og:* meta tags */
function detectMissingMeta(
  lines: string[],
  filePath?: string
): CSSLayoutFinding[] {
  try {
    const findings: CSSLayoutFinding[] = [];
    const content = lines.join('\n');
    // Only flag actual page files, not component files
    const isPage =
      /generateMetadata|export\s+(?:const|function)\s+\w*(?:Page|Layout)\b|<Head\b|<head\b|export\s+const\s+metadata\b/i.test(
        content
      );
    // Only Next.js page/layout files and explicit page components
    const isPageFile = filePath
      ? /(?:^|\/)(?:page|layout)\.[jt]sx?$/.test(filePath)
      : false;
    if (!isPage && !isPageFile) return findings;

    const hasTitle = /<title\b|title:|metadata.*title/i.test(content);
    const hasDescription =
      /description:|meta.*description|<meta\s+name=["']description/i.test(
        content
      );
    const hasOgTitle = /og:title|openGraph.*title|ogTitle/i.test(content);
    const hasOgImage = /og:image|openGraph.*image|ogImage/i.test(content);
    const hasCanonical = /canonical|alternates/i.test(content);

    if (!hasTitle) {
      findings.push({
        rule: 'SEO_MISSING_META',
        line: 1,
        message:
          'Page missing <title> or metadata.title. Critical for SEO and AI citability.',
        level: 'warning',
      });
    }
    if (!hasDescription) {
      findings.push({
        rule: 'SEO_MISSING_META',
        line: 1,
        message:
          'Page missing meta description. Important for search snippets and AI summaries.',
        level: 'warning',
      });
    }
    if (!hasOgTitle || !hasOgImage) {
      findings.push({
        rule: 'SEO_MISSING_META',
        line: 1,
        message: `Page missing ${!hasOgTitle ? 'og:title' : ''}${
          !hasOgTitle && !hasOgImage ? ' and ' : ''
        }${
          !hasOgImage ? 'og:image' : ''
        }. Required for social cards and AI citations.`,
        level: 'hint',
      });
    }
    if (!hasCanonical) {
      findings.push({
        rule: 'SEO_MISSING_META',
        line: 1,
        message:
          'Page missing canonical URL. Prevents duplicate content issues and improves AI source attribution.',
        level: 'hint',
      });
    }
    return findings;
  } catch (error) {
    throw new Error(
      `[detectMissingMeta] ${
        error instanceof Error ? error.message : String(error)
      }`
    );
  }
}

/** SEO_MISSING_STRUCTURED_DATA: pages should have JSON-LD schema.org microdata */
function detectMissingStructuredData(
  lines: string[],
  filePath?: string
): CSSLayoutFinding[] {
  try {
    const findings: CSSLayoutFinding[] = [];
    const content = lines.join('\n');
    const isPage =
      /generateMetadata|export\s+(?:const|function)\s+\w*(?:Page|Layout)\b|<Head\b|export\s+const\s+metadata\b/i.test(
        content
      );
    const isPageFile = filePath
      ? /(?:page|layout|index)\.[jt]sx?$/.test(filePath)
      : false;
    if (!isPage && !isPageFile) return findings;

    const declarations = extractStructuredDeclarations(
      content,
      filePath ?? 'inline.tsx'
    );
    if (declarations.length > 0) {
      const compiled = compileStructuredGraph(declarations, {
        routeId: filePath ?? '/',
      });
      for (const issue of compiled.issues) {
        const line = issue.sourceSpan?.startLine ?? 1;
        const severity =
          issue.severity === 'error' ? 'warning' : issue.severity;
        findings.push({
          rule: 'SEO_MISSING_STRUCTURED_DATA',
          line,
          message: `Structured-data declaration issue (${issue.code}): ${issue.message}`,
          level: severity,
        });
      }
      return findings;
    }

    const hasJsonLd =
      /application\/ld\+json|jsonLd|structuredData|schema\.org/i.test(content);
    const hasMicrodata = /itemscope|itemtype|itemprop|data-sd-type/i.test(
      content
    );
    const hasRdfa = /typeof=|property=["']og:|property=["'][^"']+["']/i.test(
      content
    );

    if (!hasJsonLd && !hasMicrodata && !hasRdfa) {
      findings.push({
        rule: 'SEO_MISSING_STRUCTURED_DATA',
        line: 1,
        message:
          'Page has no structured data (JSON-LD, microdata, or RDFa). Add schema.org markup for rich results, knowledge panels, and AI RAG/GAO citability.',
        level: 'hint',
      });
    }
    return findings;
  } catch (error) {
    throw new Error(
      `[detectMissingStructuredData] ${
        error instanceof Error ? error.message : String(error)
      }`
    );
  }
}

// ═══════════════════════════════════════════════════════════════════════
// Accessibility Detectors
// ═══════════════════════════════════════════════════════════════════════

/** A11Y_MISSING_ALT: images must have alt text */
function detectMissingAlt(lines: string[]): CSSLayoutFinding[] {
  try {
    const findings: CSSLayoutFinding[] = [];
    for (let i = 0; i < lines.length; i++) {
      const line = lines[i]!;
      if (/<img\b/.test(line) && !/alt=/.test(line)) {
        // Check next several lines for alt= (JSX img tags often span 10+ lines)
        const nearby = lines.slice(i, Math.min(i + 12, lines.length)).join(' ');
        if (!/alt=/.test(nearby) && !/role=["']presentation/.test(nearby)) {
          findings.push({
            rule: 'A11Y_MISSING_ALT',
            line: i + 1,
            message:
              '<img> without alt text. Required for screen readers (WCAG 1.1.1). Add alt="" for decorative images.',
            level: 'warning',
          });
        }
      }
    }
    return findings;
  } catch (error) {
    throw new Error(
      `[detectMissingAlt] ${
        error instanceof Error ? error.message : String(error)
      }`
    );
  }
}

/** A11Y_MISSING_LABEL: interactive elements need accessible names */
function detectMissingLabel(lines: string[]): CSSLayoutFinding[] {
  try {
    const findings: CSSLayoutFinding[] = [];
    for (let i = 0; i < lines.length; i++) {
      const line = lines[i]!;
      // Buttons with only icon children (no text content)
      if (/<button\b/.test(line)) {
        const nearby = lines.slice(i, Math.min(i + 5, lines.length)).join(' ');
        const hasLabel = /aria-label|aria-labelledby|title=|>\s*\w/.test(
          nearby
        );
        const iconOnly =
          /<(?:svg|Icon|[A-Z]\w*Icon)\b/.test(nearby) &&
          !/>[\w\s]+</.test(nearby);
        if (!hasLabel && iconOnly) {
          findings.push({
            rule: 'A11Y_MISSING_LABEL',
            line: i + 1,
            message:
              'Icon-only <button> without aria-label. Screen readers cannot describe this control (WCAG 4.1.2).',
            level: 'warning',
          });
        }
      }
      // Inputs without labels
      if (
        /<input\b/.test(line) &&
        !/type=["'](?:hidden|submit|button|reset)/.test(line)
      ) {
        // TSX often places aria-label several lines below `<input` — keep a wide forward window.
        const nearby = lines
          .slice(Math.max(0, i - 40), Math.min(i + 32, lines.length))
          .join(' ');
        if (
          !/aria-label|aria-labelledby|<label|htmlFor|placeholder/.test(nearby)
        ) {
          findings.push({
            rule: 'A11Y_MISSING_LABEL',
            line: i + 1,
            message:
              '<input> without associated label or aria-label (WCAG 1.3.1).',
            level: 'warning',
          });
        }
      }
    }
    return findings;
  } catch (error) {
    throw new Error(
      `[detectMissingLabel] ${
        error instanceof Error ? error.message : String(error)
      }`
    );
  }
}

/** A11Y_COLOR_CONTRAST: detect hardcoded low-contrast text color patterns */
function detectColorContrast(lines: string[]): CSSLayoutFinding[] {
  try {
    const findings: CSSLayoutFinding[] = [];
    // Flag very light text on light backgrounds or very dark on dark
    // Only flag contrast issues when there's evidence of a light background
    // Our codebase is dark-first (aeon tokens), so light text on dark is correct
    for (let i = 0; i < lines.length; i++) {
      const line = lines[i]!;
      // Flag dark text on what appears to be a dark background
      const darkTextOnDark =
        /(?:text-gray-[6-9]|text-neutral-[6-9]|text-slate-[6-9])/.test(line) &&
        /bg-(?:black|gray-[89]|neutral-[89]|slate-[89]|zinc-[89])/.test(line);
      // Flag very light text on light backgrounds
      const lightTextOnLight =
        /(?:text-gray-[1-2]|text-neutral-[1-2]|text-white)/.test(line) &&
        /bg-(?:white|gray-[1-2]|neutral-[1-2]|slate-[1-2])/.test(line);
      if (darkTextOnDark || lightTextOnLight) {
        findings.push({
          rule: 'A11Y_COLOR_CONTRAST',
          line: i + 1,
          message: `Potential low-contrast text: ${
            darkTextOnDark
              ? 'dark text on dark background'
              : 'light text on light background'
          }. Verify 4.5:1 ratio (WCAG AA).`,
          level: 'hint',
        });
      }
    }
    return findings;
  } catch (error) {
    throw new Error(
      `[detectColorContrast] ${
        error instanceof Error ? error.message : String(error)
      }`
    );
  }
}

// ═══════════════════════════════════════════════════════════════════════
// Boundary Violation Detectors
// Things that shouldn't be where they are.
// ═══════════════════════════════════════════════════════════════════════

// Server-side packages that should never appear in client/browser code
const SERVER_ONLY_IMPORTS = [
  'fs',
  'path',
  'child_process',
  'crypto',
  'net',
  'http',
  'https',
  'tls',
  'dns',
  'os',
  'cluster',
  'worker_threads',
  'stream',
  'readline',
  'pg',
  'mysql2',
  'mongodb',
  'redis',
  'ioredis',
  'prisma',
  'firebase-admin',
  '@google-cloud',
  'aws-sdk',
  '@aws-sdk',
  'nodemailer',
  'bcrypt',
  'jsonwebtoken',
];

// Client-side packages that shouldn't be in server/API code
// const _CLIENT_ONLY_IMPORTS = [
//   'react-dom/client',
//   'react-native',
//   '@react-native',
//   'localStorage',
//   'sessionStorage',
//   'window.',
//   'document.',
// ];

/** BOUNDARY_SERVER_IMPORT_IN_CLIENT: server modules imported in client code */
function detectServerImportInClient(
  lines: string[],
  filePath?: string
): CSSLayoutFinding[] {
  try {
    const findings: CSSLayoutFinding[] = [];
    if (!filePath) return findings;
    // Is this a client file?
    const isClient =
      /\/client\/|\.client\.|use client|'use client'|"use client"/.test(
        lines.slice(0, 10).join('\n')
      ) || /\/components\/|\/hooks\/|\.tsx$/.test(filePath);
    const isExplicitServer =
      /use server|'use server'|"use server"|\/server\/|\/api\/|\.server\./.test(
        lines.slice(0, 10).join('\n') + (filePath ?? '')
      );
    if (!isClient || isExplicitServer) return findings;

    for (let i = 0; i < Math.min(lines.length, 50); i++) {
      const line = lines[i]!;
      if (!/import\s|require\s*\(/.test(line)) continue;
      for (const pkg of SERVER_ONLY_IMPORTS) {
        if (
          line.includes(`'${pkg}'`) ||
          line.includes(`"${pkg}"`) ||
          line.includes(`from '${pkg}/`) ||
          line.includes(`from "${pkg}/`)
        ) {
          findings.push({
            rule: 'BOUNDARY_SERVER_IMPORT_IN_CLIENT',
            line: i + 1,
            message: `Server-only module '${pkg}' imported in client code. This will crash in the browser. Move to a server component or API route.`,
            level: 'error',
            crossingNumber: 1,
          });
        }
      }
    }
    return findings;
  } catch (error) {
    throw new Error(
      `[detectServerImportInClient] ${
        error instanceof Error ? error.message : String(error)
      }`
    );
  }
}

/** BOUNDARY_SECRET_IN_CLIENT: secrets/env vars that shouldn't be in client bundles */
function detectSecretInClient(
  lines: string[],
  filePath?: string
): CSSLayoutFinding[] {
  try {
    const findings: CSSLayoutFinding[] = [];
    if (!filePath) return findings;
    const isClient =
      /\/client\/|\.client\.|use client|\/components\/|\.tsx$/.test(
        lines.slice(0, 10).join('\n') + filePath
      );
    if (!isClient) return findings;

    for (let i = 0; i < lines.length; i++) {
      const line = lines[i]!;
      // Server env vars without NEXT_PUBLIC_ prefix
      if (
        /process\.env\.(?!NEXT_PUBLIC_|NODE_ENV|VERCEL_|CI\b)(\w+)/.test(line)
      ) {
        const match = line.match(/process\.env\.(\w+)/);
        if (match && !/NEXT_PUBLIC_/.test(match[1]!)) {
          findings.push({
            rule: 'BOUNDARY_SECRET_IN_CLIENT',
            line: i + 1,
            message: `Server env var 'process.env.${match[1]}' in client code. Only NEXT_PUBLIC_* vars are available in the browser. This will be undefined at runtime.`,
            level: 'error',
            crossingNumber: 1,
          });
        }
      }
      // Hardcoded API keys/tokens
      if (
        /(?:api[_-]?key|secret|token|password|auth)\s*[:=]\s*["'][A-Za-z0-9_-]{20,}["']/i.test(
          line
        )
      ) {
        findings.push({
          rule: 'BOUNDARY_SECRET_IN_CLIENT',
          line: i + 1,
          message:
            'Possible hardcoded secret in source code. Use environment variables instead.',
          level: 'error',
          crossingNumber: 1,
        });
      }
    }
    return findings;
  } catch (error) {
    throw new Error(
      `[detectSecretInClient] ${
        error instanceof Error ? error.message : String(error)
      }`
    );
  }
}

// ═══════════════════════════════════════════════════════════════════════
// Caching Footgun Detectors
// Things that shoot your foot off when you cannot cache-purge.
// ═══════════════════════════════════════════════════════════════════════

/** CACHE_IMMUTABLE_MUTABLE_RESOURCE: long-lived caching on mutable content */
function detectCacheFootguns(
  lines: string[],
  _filePath?: string
): CSSLayoutFinding[] {
  try {
    const findings: CSSLayoutFinding[] = [];
    for (let i = 0; i < lines.length; i++) {
      const line = lines[i]!;
      // HSTS / unrelated security headers reuse max-age= — not HTTP cache semantics.
      if (/Strict-Transport-Security/i.test(line)) {
        continue;
      }
      // max-age very high without content hash in URL
      if (/max-age\s*=\s*(\d+)/.test(line)) {
        const match = line.match(/max-age\s*=\s*(\d+)/);
        const maxAge = parseInt(match?.[1] ?? '0', 10);
        if (maxAge > 86400 * 30) {
          // Long-lived cache entries are safe only when the surrounding branch
          // proves a content-addressed asset URL.
          const ctx = lines
            .slice(Math.max(0, i - 25), Math.min(i + 1, lines.length))
            .join('\n');
          const hasHash =
            /\[hash\]|\[contenthash\]|\.[\da-f]{8,}\.|\/assets\/[^/]+-[a-zA-Z0-9]{6,}\./i.test(
              ctx
            ) ||
            (/\.test\(resolution\.candidatePath\)/.test(ctx) &&
              /\bassets\b/.test(ctx));
          if (!hasHash) {
            findings.push({
              rule: 'CACHE_IMMUTABLE_MUTABLE_RESOURCE',
              line: i + 1,
              message: `Cache max-age=${maxAge} (${Math.round(
                maxAge / 86400
              )} days) without content hash. If this resource changes, users will be stuck with the stale version. Use content-hashed filenames or lower max-age.`,
              level: 'warning',
              crossingNumber: 1,
            });
          }
        }
      }
      // the immutability directive requires a content-addressed URL
      if (
        /immutable/.test(line) &&
        /Cache-Control|cache-control|cacheControl/.test(line)
      ) {
        const nearby = lines
          .slice(Math.max(0, i - 25), Math.min(i + 3, lines.length))
          .join(' ');
        if (
          !/\[hash\]|\[contenthash\]|\.[\da-f]{8,}\.|\/assets\/[^/]+-[a-zA-Z0-9]{6,}\./i.test(
            nearby
          ) &&
          !(
            /\.test\(resolution\.candidatePath\)/.test(nearby) &&
            /\bassets\b/.test(nearby)
          )
        ) {
          findings.push({
            rule: 'CACHE_IMMUTABLE_MUTABLE_RESOURCE',
            line: i + 1,
            message:
              'Long-lived immutable caching without content-hashed URL. This resource can never be updated for existing users without a URL change.',
            level: 'warning',
            crossingNumber: 1,
          });
        }
      }
    }
    return findings;
  } catch (error) {
    throw new Error(
      `[detectCacheFootguns] ${
        error instanceof Error ? error.message : String(error)
      }`
    );
  }
}

// ═══════════════════════════════════════════════════════════════════════
// Slop Detectors
// Things that make code worse. Negative reinforcement.
// ═══════════════════════════════════════════════════════════════════════

/** SLOP_EMOJI_IN_CODE: emojis in user-facing strings (optional) */
function detectEmojiInCode(lines: string[]): CSSLayoutFinding[] {
  try {
    const findings: CSSLayoutFinding[] = [];
    const emojiPattern =
      // eslint-disable-next-line no-misleading-character-class
      /[\u{1F600}-\u{1F64F}\u{1F300}-\u{1F5FF}\u{1F680}-\u{1F6FF}\u{1F1E0}-\u{1F1FF}\u{2600}-\u{26FF}\u{2700}-\u{27BF}\u{FE00}-\u{FE0F}\u{1F900}-\u{1F9FF}\u{1FA00}-\u{1FA6F}\u{1FA70}-\u{1FAFF}]/u;
    for (let i = 0; i < lines.length; i++) {
      const line = lines[i]!;
      // Only flag emojis in JSX string content and string literals, not comments
      if (/^\s*\/\/|^\s*\*|^\s*\/\*/.test(line)) continue;
      if (emojiPattern.test(line)) {
        // Skip if it's in a test file or story
        const match = line.match(emojiPattern);
        findings.push({
          rule: 'SLOP_EMOJI_IN_CODE',
          line: i + 1,
          message: `Emoji '${match?.[0]}' in source code. Use text or icons instead for accessibility and professionalism.`,
          level: 'hint',
        });
      }
    }
    return findings;
  } catch (error) {
    throw new Error(
      `[detectEmojiInCode] ${
        error instanceof Error ? error.message : String(error)
      }`
    );
  }
}

/** SLOP_CONSOLE_LOG: console.log left in production code */
function detectConsoleLog(
  lines: string[],
  filePath?: string
): CSSLayoutFinding[] {
  try {
    const findings: CSSLayoutFinding[] = [];
    // Skip test files and scripts
    if (filePath && /\.test\.|\.spec\.|__tests__|scripts\//.test(filePath))
      return findings;
    for (let i = 0; i < lines.length; i++) {
      const line = lines[i]!;
      if (/^\s*\/\//.test(line)) continue; // skip comments
      if (/\bconsole\.log\s*\(/.test(line)) {
        const trimmed = line.trimStart();
        const isSingleLine = /\bconsole\.log\s*\(.*\)\s*;?$/.test(line);

        const finding: CSSLayoutFinding = {
          rule: 'SLOP_CONSOLE_LOG_IN_PROD',
          line: i + 1,
          message:
            'console.log() in production code. Use a structured logger or remove before shipping.',
          level: 'hint',
          ...(isSingleLine
            ? {
                fix: {
                  line: i + 1,
                  oldText: trimmed,
                  newText: `// ${trimmed}`,
                  description: 'Comment out console.log',
                },
              }
            : {}),
        };
        findings.push(finding);
      }
    }
    return findings;
  } catch (error) {
    throw new Error(
      `[detectConsoleLog] ${
        error instanceof Error ? error.message : String(error)
      }`
    );
  }
}

/** SLOP_TODO: unresolved TODO/FIXME/HACK/XXX comments */
function detectTodoComments(lines: string[]): CSSLayoutFinding[] {
  try {
    const findings: CSSLayoutFinding[] = [];
    for (let i = 0; i < lines.length; i++) {
      const line = lines[i]!;
      const match = line.match(/\b(TODO|FIXME|HACK|XXX)\b[:\s]*(.*)/);
      if (match) {
        findings.push({
          rule: 'SLOP_TODO_IN_PROD',
          line: i + 1,
          message: `Unresolved ${match[1]}: ${
            (match[2] ?? '').trim().slice(0, 60) || '(no description)'
          }`,
          level: 'hint',
        });
      }
    }
    return findings;
  } catch (error) {
    throw new Error(
      `[detectTodoComments] ${
        error instanceof Error ? error.message : String(error)
      }`
    );
  }
}

/** SLOP_HARDCODED_URL: hardcoded localhost/staging URLs in production code */
function detectHardcodedUrls(
  lines: string[],
  filePath?: string
): CSSLayoutFinding[] {
  try {
    const findings: CSSLayoutFinding[] = [];
    if (
      filePath &&
      (/\.test\.|\.spec\.|__tests__|\.stories\.|\.config\./.test(filePath) ||
        // Local puppeteer harnesses (docs-app): URLs come from env; README holds examples.
        /[/\\]puppeteer-test[^/\\]*\.js$/.test(filePath) ||
        // Dev-only defaults: DEV_LOCAL_HTTP_HOST; literals are localhost fallbacks.
        /[/\\]dev-default-origin\.ts$/.test(filePath) ||
        // Zedge companion is a local sidecar; loopback URLs are its API contract.
        /(?:^|[/\\])open-source[/\\]zedge[/\\]companion[/\\]src[/\\]/.test(
          filePath
        ) ||
        // Self-tickle / Node shim loopback in GCP coordinators (same process only).
        /[/\\]gcp-coordinator\.ts$/.test(filePath) ||
        /[/\\]gcp-generic-coordinator\.ts$/.test(filePath) ||
        // Jest/Vitest bootstrap often sets loopback fixtures for integration tests.
        /[/\\]jest\.setup[^/\\]*\.(js|ts)$/.test(filePath))
    )
      return findings;
    for (let i = 0; i < lines.length; i++) {
      const line = lines[i]!;
      if (/^\s*\/\//.test(line)) continue;
      if (/https?:\/\/localhost[:\d/]|https?:\/\/127\.0\.0\.1/.test(line)) {
        findings.push({
          rule: 'SLOP_HARDCODED_URL',
          line: i + 1,
          message:
            'Hardcoded localhost URL. Use environment variable for base URL.',
          level: 'warning',
        });
      }
      if (
        /https?:\/\/.*staging\.|https?:\/\/.*\.internal\./.test(line) &&
        !/\/\/\s/.test(line)
      ) {
        findings.push({
          rule: 'SLOP_HARDCODED_URL',
          line: i + 1,
          message:
            'Hardcoded staging/internal URL in production code. Use environment variable.',
          level: 'warning',
        });
      }
    }
    return findings;
  } catch (error) {
    throw new Error(
      `[detectHardcodedUrls] ${
        error instanceof Error ? error.message : String(error)
      }`
    );
  }
}

// ── Public API ───────────────────────────────────────────────────────

/**
 * Run all static analysis detectors on JSX/TSX source text.
 * Covers: CSS layout, Tailwind conflicts, SEO, structured data,
 * accessibility, boundary violations, caching, and slop.
 * Returns findings, diagnostics (for Betty), and obligations (for SemanticContract).
 */
export function analyzeCSSLayout(
  sourceText: string,
  filePath?: string
): CSSLayoutAnalysisResult {
  try {
    const ext = filePath
      ? filePath.slice(filePath.lastIndexOf('.')).toLowerCase()
      : '';
    if (
      ext &&
      ext !== '.tsx' &&
      ext !== '.jsx' &&
      ext !== '.ts' &&
      ext !== '.js'
    ) {
      return { findings: [], diagnostics: [], obligations: [] };
    }

    const lines = sourceText.split('\n');
    const findings: CSSLayoutFinding[] = [
      // CSS Layout Thermodynamics
      ...detectTailwindConflicts(lines),
      ...detectFlexOverflow(lines),
      ...detectPositionEscape(lines),
      ...detectResponsiveNonMonotone(lines),
      ...detectMarginCollapse(lines),
      ...detectMissingFlexParent(lines),
      ...detectOverflowUnhandled(lines),
      // SEO / Structured Data / AI Citability
      ...detectHeadingHierarchy(lines),
      ...detectMissingMeta(lines, filePath),
      ...detectMissingStructuredData(lines, filePath),
      // Accessibility
      ...detectMissingAlt(lines),
      ...detectMissingLabel(lines),
      ...detectColorContrast(lines),
      // Boundary violations
      ...detectServerImportInClient(lines, filePath),
      ...detectSecretInClient(lines, filePath),
      // Caching footguns
      ...detectCacheFootguns(lines, filePath),
      // Slop (negative reinforcement)
      ...detectEmojiInCode(lines),
      ...detectConsoleLog(lines, filePath),
      ...detectTodoComments(lines),
      ...detectHardcodedUrls(lines, filePath),
    ];

    return {
      findings,
      diagnostics: findings.map(findingToDiagnostic),
      obligations: findings.map(findingToObligation),
    };
  } catch (error) {
    throw new Error(
      `[analyzeCSSLayout] ${
        error instanceof Error ? error.message : String(error)
      }`
    );
  }
}

/** List of all diagnostic codes for filtering. */
export const CSS_LAYOUT_DIAGNOSTIC_CODES: readonly DiagnosticCode[] = [
  'LAYOUT_TAILWIND_CONFLICT',
  'LAYOUT_FLEX_OVERFLOW',
  'LAYOUT_POSITION_ESCAPE',
  'LAYOUT_RESPONSIVE_NON_MONOTONE',
  'LAYOUT_MARGIN_COLLAPSE',
  'LAYOUT_MISSING_FLEX_PARENT',
  'LAYOUT_OVERFLOW_UNHANDLED',
  'SEO_HEADING_HIERARCHY',
  'SEO_MISSING_META',
  'SEO_MISSING_STRUCTURED_DATA',
  'A11Y_MISSING_ALT',
  'A11Y_MISSING_LABEL',
  'A11Y_COLOR_CONTRAST',
  'A11Y_MISSING_LANG',
  'BOUNDARY_SERVER_IMPORT_IN_CLIENT',
  'BOUNDARY_CLIENT_IMPORT_IN_SERVER',
  'BOUNDARY_SECRET_IN_CLIENT',
  'CACHE_IMMUTABLE_MUTABLE_RESOURCE',
  'CACHE_NO_CACHE_HEADERS',
  'SLOP_EMOJI_IN_CODE',
  'SLOP_CONSOLE_LOG_IN_PROD',
  'SLOP_TODO_IN_PROD',
  'SLOP_HARDCODED_URL',
];
