// @a0n/.ignored_gnosis/src/dark-shroud.ts
/**
 * Dark Shroud
 *
 * This reducer is theorem-informed, not theorem-proved.
 *
 * It reuses the existing finite-state mesh/Buleyean theorem family as a
 * design constraint: when it emits an inspection floor, it uses the same
 * `1 / (n * (R + 1))` lower-bound shape proved for Buleyean-normalized mesh
 * kernels in Lean. But this file does not itself construct that kernel in Lean
 * or prove that every route-decision branch realizes the same contraction.
 *
 * Honest status:
 * - backed by existing Lean results for the floor formula and contraction at
 *   the abstract mesh-kernel level
 * - not yet a new Lean mechanization of port-route policy synthesis
 * - intended as the runtime contract that a future Lean `PortDomainMesh` or
 *   `DarkShroudKernel` module should refine
 */

import {
  auditXRangerBoundaryManifest,
  createXRangerBoundaryManifest,
  type XRangerBoundaryAction,
  type XRangerBoundaryManifest,
  type XRangerBoundaryManifestAudit,
  type XRangerDirection,
  type XRangerInspectionAction,
  type XRangerInspectionFloor,
  type XRangerPrismShroud,
  type XRangerStateClass,
  type XRangerTransport,
  type XRangerTrustTier,
} from './x-ranger-boundary';
import { Verdict } from '@a0n/bitwise/triton-verdict';
import { sha256Hex, stableStringify } from './universe/index';

export const DARK_SHROUD_THEOREM_IDS = [
  'THM-MESH-MASKED-SOFTMAX-LINF-NONEXPANSIVE',
  'THM-MESH-DOBRUSHIN-OSC-CONTRACTION',
  'THM-BULEYEAN-DIST-UNIFORM-FLOOR',
  'THM-BULEYEAN-KERNEL-DOBRUSHIN-FLOOR',
  'THM-MESH-DISTRIBUTED-CERT-GLOBALIZES',
  'THM-MESH-OBSERVABILITY-COMPLETION-MONOTONE',
] as const;

export type DarkShroudTheoremId =
  (typeof DARK_SHROUD_THEOREM_IDS)[number];

export type DarkShroudEvidenceMode = 'none' | 'mirror' | 'snapshot';

export type DarkShroudShroudMode = XRangerPrismShroud['mode'];

export interface DarkShroudRouteInput {
  readonly routeId: string;
  readonly service: string;
  readonly localPort: number;
  readonly remotePortClass: string;
  readonly protocol: XRangerTransport;
  readonly direction: XRangerDirection;
  readonly trustTier: XRangerTrustTier;
  readonly riskScore: number;
  readonly preserveConnectivity?: boolean;
  readonly evidenceMode?: DarkShroudEvidenceMode;
  readonly shroudMode?: DarkShroudShroudMode;
}

export interface DeriveDarkShroudInput {
  readonly generatedAt?: string;
  readonly source?: string;
  readonly rejectionBudget?: number;
  readonly routes: readonly DarkShroudRouteInput[];
}

export interface DarkShroudDecision {
  readonly routeId: string;
  readonly service: string;
  readonly stateClassId: string;
  readonly localPort: number;
  readonly primaryAction: XRangerBoundaryAction;
  readonly secondaryActions: readonly XRangerBoundaryAction[];
  readonly shroudMode: DarkShroudShroudMode;
  readonly inspectionFloor: number | null;
  readonly reason: string;
  /**
   * The canonical ternary verdict (`@a0n/bitwise`) derived from
   * `primaryAction`. Additive: the string disposition above is the lossy,
   * back-compat form; this field carries the signed-decision-with-neutral the
   * grid needs. `allow` → Accept, `quarantine` → Abstain (the neutral middle the
   * string already latently expresses), hard `drop`/shroud → Decline. The
   * inspect-and-admit actions (`mirror`/`rate_limit`/`snapshot`) keep the route
   * admissible under observation, so they map to Accept.
   */
  readonly verdict: Verdict;
}

export interface DarkShroudBundle {
  readonly kind: 'gnosis-dark-shroud';
  readonly version: 1;
  readonly generatedAt: string;
  readonly source: string;
  readonly digest: string;
  readonly theoremIds: readonly DarkShroudTheoremId[];
  readonly rejectionBudget: number;
  readonly stateClassCount: number;
  readonly weakestInspectionFloor: number | null;
  readonly decisions: readonly DarkShroudDecision[];
  readonly xRangerManifest: XRangerBoundaryManifest;
}

export interface DarkShroudBundleAudit {
  readonly selfConsistent: boolean;
  readonly issues: readonly string[];
  readonly boundaryAudit: XRangerBoundaryManifestAudit;
}

interface NormalizedRoute {
  readonly routeId: string;
  readonly service: string;
  readonly localPort: number;
  readonly remotePortClass: string;
  readonly protocol: XRangerTransport;
  readonly direction: XRangerDirection;
  readonly trustTier: XRangerTrustTier;
  readonly riskScore: number;
  readonly preserveConnectivity: boolean;
  readonly evidenceMode: DarkShroudEvidenceMode;
  readonly shroudMode: DarkShroudShroudMode;
}

interface DecisionShape {
  readonly primaryAction: XRangerBoundaryAction;
  readonly secondaryActions: readonly XRangerBoundaryAction[];
  readonly shroudMode: DarkShroudShroudMode;
  readonly reason: string;
}

const INSPECTION_ACTIONS = new Set<XRangerInspectionAction>([
  'mirror',
  'quarantine',
  'snapshot',
]);

function nowIso(): string {
  return new Date().toISOString();
}

function clampPort(value: number): number {
  if (!Number.isFinite(value)) {
    return 0;
  }
  return Math.max(0, Math.min(65535, Math.floor(value)));
}

function clampRiskScore(value: number): number {
  if (!Number.isFinite(value)) {
    return 0;
  }
  return Math.max(0, Math.min(100, Math.floor(value)));
}

function clampRejectionBudget(value: number | undefined): number {
  if (value === undefined || !Number.isFinite(value)) {
    return 7;
  }
  return Math.max(0, Math.floor(value));
}

function trimOrFallback(value: string, fallback: string): string {
  const normalized = value.trim();
  return normalized.length > 0 ? normalized : fallback;
}

function uniqueSortedActions(
  actions: readonly XRangerBoundaryAction[]
): XRangerBoundaryAction[] {
  return [...new Set(actions)].sort((left, right) => left.localeCompare(right));
}

function uniqueSortedInspectionActions(
  actions: readonly XRangerInspectionAction[]
): XRangerInspectionAction[] {
  return [...new Set(actions)].sort((left, right) => left.localeCompare(right));
}

function stateClassIdForRoute(route: NormalizedRoute): string {
  return [
    route.service,
    `p${route.localPort}`,
    route.remotePortClass,
    route.protocol,
    route.direction,
    route.trustTier,
  ].join(':');
}

function normalizeRoute(route: DarkShroudRouteInput): NormalizedRoute {
  return {
    routeId: trimOrFallback(route.routeId, 'route'),
    service: trimOrFallback(route.service, 'service'),
    localPort: clampPort(route.localPort),
    remotePortClass: trimOrFallback(route.remotePortClass, 'any'),
    protocol: route.protocol,
    direction: route.direction,
    trustTier: route.trustTier,
    riskScore: clampRiskScore(route.riskScore),
    preserveConnectivity: route.preserveConnectivity ?? false,
    evidenceMode: route.evidenceMode ?? 'none',
    shroudMode: route.shroudMode ?? 'none',
  };
}

function compareRoutes(left: NormalizedRoute, right: NormalizedRoute): number {
  return left.routeId.localeCompare(right.routeId);
}

function compareStateClasses(left: XRangerStateClass, right: XRangerStateClass): number {
  return left.id.localeCompare(right.id);
}

function routeToStateClass(route: NormalizedRoute): XRangerStateClass {
  return {
    id: stateClassIdForRoute(route),
    localPortClass: `${route.service}:${route.localPort}`,
    remotePortClass: route.remotePortClass,
    protocol: route.protocol,
    direction: route.direction,
    trustTier: route.trustTier,
  };
}

function pickDefaultShroudMode(route: NormalizedRoute): DarkShroudShroudMode {
  if (route.trustTier === 'control-plane' || route.trustTier === 'privileged') {
    return 'attested-route';
  }
  if (route.riskScore >= 70) {
    return 'delayed-attested-route';
  }
  if (route.evidenceMode !== 'none') {
    return 'attested-route';
  }
  return 'none';
}

function normalizeShroudMode(route: NormalizedRoute): DarkShroudShroudMode {
  return route.shroudMode === 'none' ? pickDefaultShroudMode(route) : route.shroudMode;
}

function decideRoute(route: NormalizedRoute): DecisionShape {
  const shroudMode = normalizeShroudMode(route);

  if (route.riskScore >= 85 && !route.preserveConnectivity) {
    return {
      primaryAction: 'quarantine',
      secondaryActions: ['snapshot'],
      shroudMode: shroudMode === 'none' ? 'attested-route' : shroudMode,
      reason: 'critical route risk exceeded containment threshold',
    };
  }

  if (route.riskScore >= 70) {
    return {
      primaryAction: 'mirror',
      secondaryActions: uniqueSortedActions(['rate_limit', route.evidenceMode === 'snapshot' ? 'snapshot' : 'mirror']).filter(
        (action) => action !== 'mirror'
      ),
      shroudMode: shroudMode === 'none' ? 'delayed-attested-route' : shroudMode,
      reason: 'elevated route risk requires mirrored inspection with bounded pressure',
    };
  }

  if (route.evidenceMode === 'snapshot') {
    return {
      primaryAction: 'mirror',
      secondaryActions: ['snapshot'],
      shroudMode: shroudMode === 'none' ? 'attested-route' : shroudMode,
      reason: 'explicit evidence capture requested for this route',
    };
  }

  if (
    route.evidenceMode === 'mirror' ||
    route.trustTier === 'control-plane' ||
    route.trustTier === 'privileged'
  ) {
    return {
      primaryAction: 'mirror',
      secondaryActions: [],
      shroudMode: shroudMode === 'none' ? 'attested-route' : shroudMode,
      reason: 'privileged or evidence-bound route kept under mirrored observation',
    };
  }

  if (route.riskScore >= 40) {
    return {
      primaryAction: 'mirror',
      secondaryActions: ['rate_limit'],
      shroudMode,
      reason: 'moderate route risk receives inspection before open admission',
    };
  }

  return {
    primaryAction: 'allow',
    secondaryActions: [],
    shroudMode,
    reason: 'route remains admissible without intervention',
  };
}

/**
 * Derive the canonical ternary {@link Verdict} from a boundary disposition.
 * The string `primaryAction` is lossy (it cannot name the neutral middle as a
 * first-class state); the verdict recovers it:
 *   - `allow`                              → Accept  (admit, no intervention)
 *   - `quarantine`                         → Abstain (the latent neutral middle)
 *   - `drop`                               → Decline (hard reject / shroud)
 *   - `mirror` / `rate_limit` / `snapshot` → Accept  (admit under observation)
 */
export function verdictForAction(action: XRangerBoundaryAction): Verdict {
  switch (action) {
    case 'quarantine':
      return Verdict.Abstain;
    case 'drop':
      return Verdict.Decline;
    case 'allow':
    case 'mirror':
    case 'rate_limit':
    case 'snapshot':
      return Verdict.Accept;
  }
}

function hasInspectionAction(decision: DecisionShape): boolean {
  if (INSPECTION_ACTIONS.has(decision.primaryAction as XRangerInspectionAction)) {
    return true;
  }
  return decision.secondaryActions.some((action) =>
    INSPECTION_ACTIONS.has(action as XRangerInspectionAction)
  );
}

function ensureInspectionFloor(
  routes: readonly NormalizedRoute[],
  decisions: readonly DecisionShape[]
): DecisionShape[] {
  if (decisions.some(hasInspectionAction)) {
    return [...decisions];
  }
  if (routes.length === 0) {
    return [...decisions];
  }

  const promotedIndex = routes.reduce((bestIndex, route, index) => {
    if (bestIndex === -1) {
      return index;
    }
    const bestRoute = routes[bestIndex];
    if (!bestRoute) {
      return index;
    }
    return route.riskScore > bestRoute.riskScore ? index : bestIndex;
  }, -1);

  return decisions.map((decision, index) =>
    index === promotedIndex
      ? {
          primaryAction: 'mirror',
          secondaryActions: decision.primaryAction === 'allow' ? [] : [decision.primaryAction],
          shroudMode:
            decision.shroudMode === 'none' ? 'attested-route' : decision.shroudMode,
          reason: 'baseline dark shroud promoted one route into mirrored observation',
        }
      : decision
  );
}

function weakestInspectionFloor(
  stateClassCount: number,
  rejectionBudget: number
): number | null {
  if (stateClassCount <= 0) {
    return null;
  }
  return 1 / (stateClassCount * (rejectionBudget + 1));
}

function inspectionFloorsForActions(
  actions: readonly XRangerInspectionAction[],
  stateClassCount: number,
  rejectionBudget: number
): XRangerInspectionFloor[] {
  const denominator = stateClassCount * (rejectionBudget + 1);
  return actions.map((action) => ({
    action,
    numerator: 1,
    denominator,
  }));
}

function digestDarkShroudBody(body: Omit<DarkShroudBundle, 'digest'>): string {
  return sha256Hex(stableStringify(body));
}

/**
 * Handles the gnosis derive Dark Shroud workflow.
 */
export function deriveDarkShroud(
  input: DeriveDarkShroudInput
): DarkShroudBundle {
  const source = trimOrFallback(input.source ?? 'gnosis-dark-shroud', 'gnosis-dark-shroud');
  const generatedAt = input.generatedAt ?? nowIso();
  const rejectionBudget = clampRejectionBudget(input.rejectionBudget);
  const routes = [...input.routes].map(normalizeRoute).sort(compareRoutes);
  const preliminaryDecisions = routes.map(decideRoute);
  const decisionsWithFloor = ensureInspectionFloor(routes, preliminaryDecisions);

  const stateClasses = [...new Map(
    routes.map((route) => {
      const stateClass = routeToStateClass(route);
      return [stateClass.id, stateClass] as const;
    })
  ).values()].sort(compareStateClasses);

  const stateClassCount = stateClasses.length;
  const inspectionFloorValue = weakestInspectionFloor(stateClassCount, rejectionBudget);

  const inspectableActions = uniqueSortedInspectionActions(
    decisionsWithFloor.flatMap((decision) => {
      const actions: XRangerInspectionAction[] = [];
      if (INSPECTION_ACTIONS.has(decision.primaryAction as XRangerInspectionAction)) {
        actions.push(decision.primaryAction as XRangerInspectionAction);
      }
      for (const action of decision.secondaryActions) {
        if (INSPECTION_ACTIONS.has(action as XRangerInspectionAction)) {
          actions.push(action as XRangerInspectionAction);
        }
      }
      return actions;
    })
  );

  const manifestActions = uniqueSortedActions(
    decisionsWithFloor.flatMap((decision) => [
      decision.primaryAction,
      ...decision.secondaryActions,
    ])
  );

  const globalShroudMode = decisionsWithFloor.some(
    (decision) => decision.shroudMode === 'delayed-attested-route'
  )
    ? 'delayed-attested-route'
    : decisionsWithFloor.some((decision) => decision.shroudMode === 'attested-route')
      ? 'attested-route'
      : 'none';

  const xRangerManifest = createXRangerBoundaryManifest({
    generatedAt,
    theoremIds: [...DARK_SHROUD_THEOREM_IDS],
    policy: {
      stateClasses,
      actions: manifestActions,
      inspectionFloors: inspectionFloorsForActions(
        inspectableActions,
        stateClassCount,
        rejectionBudget
      ),
      certificationMode: 'local-quorum-globalized',
      reconstructionMode: 'spec-boundary',
      prismShroud: {
        mode: globalShroudMode,
        privacyModel: 'accountability',
      },
    },
  });

  const decisions: DarkShroudDecision[] = routes.map((route, index) => {
    const decision = decisionsWithFloor[index];
    if (!decision) {
      throw new Error(`Missing decision for route ${route.routeId}`);
    }

    return {
      routeId: route.routeId,
      service: route.service,
      stateClassId: stateClassIdForRoute(route),
      localPort: route.localPort,
      primaryAction: decision.primaryAction,
      secondaryActions: decision.secondaryActions,
      shroudMode: decision.shroudMode,
      inspectionFloor: hasInspectionAction(decision) ? inspectionFloorValue : null,
      reason: decision.reason,
      verdict: verdictForAction(decision.primaryAction),
    };
  });

  const body = {
    kind: 'gnosis-dark-shroud' as const,
    version: 1 as const,
    generatedAt,
    source,
    theoremIds: [...DARK_SHROUD_THEOREM_IDS],
    rejectionBudget,
    stateClassCount,
    weakestInspectionFloor: inspectionFloorValue,
    decisions,
    xRangerManifest,
  };

  return {
    ...body,
    digest: digestDarkShroudBody(body),
  };
}

/**
 * Handles the gnosis audit Dark Shroud Bundle workflow.
 */
export function auditDarkShroudBundle(
  bundle: DarkShroudBundle
): DarkShroudBundleAudit {
  const issues: string[] = [];
  const expectedDigest = digestDarkShroudBody({
    kind: bundle.kind,
    version: bundle.version,
    generatedAt: bundle.generatedAt,
    source: bundle.source,
    theoremIds: bundle.theoremIds,
    rejectionBudget: bundle.rejectionBudget,
    stateClassCount: bundle.stateClassCount,
    weakestInspectionFloor: bundle.weakestInspectionFloor,
    decisions: bundle.decisions,
    xRangerManifest: bundle.xRangerManifest,
  });

  if (bundle.digest !== expectedDigest) {
    issues.push('digest does not match the canonical dark shroud body');
  }

  if (bundle.decisions.length === 0) {
    issues.push('dark shroud must contain at least one route decision');
  }

  if (bundle.stateClassCount !== bundle.xRangerManifest.policy.stateClasses.length) {
    issues.push('state class count does not match x-ranger manifest policy');
  }

  const knownStateClasses = new Set(
    bundle.xRangerManifest.policy.stateClasses.map((stateClass) => stateClass.id)
  );
  for (const decision of bundle.decisions) {
    if (!knownStateClasses.has(decision.stateClassId)) {
      issues.push(`decision '${decision.routeId}' references an unknown state class`);
    }
    if (!bundle.xRangerManifest.policy.actions.includes(decision.primaryAction)) {
      issues.push(
        `decision '${decision.routeId}' primary action is missing from the x-ranger manifest`
      );
    }
    for (const secondaryAction of decision.secondaryActions) {
      if (!bundle.xRangerManifest.policy.actions.includes(secondaryAction)) {
        issues.push(
          `decision '${decision.routeId}' secondary action '${secondaryAction}' is missing from the x-ranger manifest`
        );
      }
    }
  }

  const boundaryAudit = auditXRangerBoundaryManifest(bundle.xRangerManifest);
  if (!boundaryAudit.selfConsistent) {
    issues.push('x-ranger boundary manifest audit failed');
  }

  return {
    selfConsistent: issues.length === 0 && boundaryAudit.selfConsistent,
    issues,
    boundaryAudit,
  };
}
