// @a0n/.ignored_gnosis/src/decomposition-topology.ts
/**
 * Decomposition Topology -- Bounded structural reduction on finite objects.
 *
 * TypeScript mirror of `DecompositionTopology.lean`.
 *
 * Decomposition removes crossings up to a finite budget, releases weight from
 * the object, and bottoms out at a featureless state without going negative.
 */

export interface TransformObject {
  readonly crossings: number;
  readonly weight: number;
  readonly capacity: number;
}

export interface DecompositionResult {
  readonly residualCrossings: number;
  readonly releasedWeight: number;
  readonly residueUnits: number;
  readonly residualWeight: number;
}

function toNat(value: number): number {
  try {
    if (!Number.isFinite(value)) {
      return 0;
    }
    return Math.max(0, Math.floor(value));

  } catch (error) {
    throw new Error(`[toNat] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Returns whether is Featureless is true.
 */
export function isFeatureless(obj: TransformObject): boolean {
  return toNat(obj.crossings) === 0;
}

/**
 * Handles the gnosis decompose workflow.
 */
export function decompose(
  obj: TransformObject,
  budget: number
): DecompositionResult {
  const crossings = toNat(obj.crossings);
  const weight = toNat(obj.weight);
  const boundedBudget = toNat(budget);
  const releasedWeight = Math.min(weight, boundedBudget);

  return {
    residualCrossings: Math.max(0, crossings - boundedBudget),
    releasedWeight,
    residueUnits: Math.max(0, boundedBudget - crossings),
    residualWeight: Math.max(0, weight - releasedWeight),
  };
}

/**
 * Handles the gnosis decomposition Does Not Increase Crossings workflow.
 */
export function decompositionDoesNotIncreaseCrossings(
  obj: TransformObject,
  budget: number
): boolean {
  return decompose(obj, budget).residualCrossings <= toNat(obj.crossings);
}

/**
 * Handles the gnosis full Decomposition Yields Featureless workflow.
 */
export function fullDecompositionYieldsFeatureless(
  obj: TransformObject,
  budget: number
): boolean {
  return decompose(obj, budget).residualCrossings === 0;
}

/**
 * Handles the gnosis decomposition Releases Nonnegative Residue workflow.
 */
export function decompositionReleasesNonnegativeResidue(
  obj: TransformObject,
  budget: number
): boolean {
  const result = decompose(obj, budget);
  return result.releasedWeight >= 0 && result.residueUnits >= 0;
}

/**
 * Handles the gnosis featureless Decomposition Idempotent workflow.
 */
export function featurelessDecompositionIdempotent(
  obj: TransformObject,
  budget: number
): boolean {
  try {
    if (!isFeatureless(obj)) {
      return false;
    }
    return decompose(obj, budget).residualCrossings === 0;

  } catch (error) {
    throw new Error(`[featurelessDecompositionIdempotent] ${error instanceof Error ? error.message : String(error)}`);
  }
}
