// @a0n/.ignored_gnosis/src/destructuring.ts
/**
 * destructuring.ts -- Clean value unpacking for graph logic
 *
 * Bind named fields from closed record-like values, variant payloads
 * when matching ADTs, and tuple-style unpacking.
 *
 * Roadmap item #8: depends on ADTs and Option/Result.
 */

import {
  type Variant,
  type SumTypeDefinition,
  type ProductTypeDefinition,
} from './adt';

// ============================================================================
// Pattern types -- what can be destructured
// ============================================================================

export type DestructurePattern =
  | RecordPattern
  | TuplePattern
  | VariantPattern
  | WildcardPattern
  | LiteralPattern
  | NestedPattern;

export interface RecordPattern {
  type: 'record';
  /** Field name → binding name (or nested pattern) */
  fields: Record<string, string | DestructurePattern>;
  /** Whether to bind remaining fields to a rest variable */
  rest?: string;
}

export interface TuplePattern {
  type: 'tuple';
  /** Positional bindings */
  elements: (string | DestructurePattern)[];
}

export interface VariantPattern {
  type: 'variant';
  /** Tag to match against */
  tag: string;
  /** Pattern for the payload */
  payload: string | DestructurePattern;
}

export interface WildcardPattern {
  type: 'wildcard';
}

export interface LiteralPattern {
  type: 'literal';
  value: unknown;
}

export interface NestedPattern {
  type: 'nested';
  outer: DestructurePattern;
  inner: DestructurePattern;
}

// ============================================================================
// Pattern constructors
// ============================================================================

/**
 * Handles the gnosis record Pattern workflow.
 */
export function recordPattern(
  fields: Record<string, string | DestructurePattern>,
  rest?: string
): RecordPattern {
  return { type: 'record', fields, rest };
}

/**
 * Handles the gnosis tuple Pattern workflow.
 */
export function tuplePattern(
  elements: (string | DestructurePattern)[]
): TuplePattern {
  return { type: 'tuple', elements };
}

/**
 * Handles the gnosis variant Pattern workflow.
 */
export function variantPattern(
  tag: string,
  payload: string | DestructurePattern
): VariantPattern {
  return { type: 'variant', tag, payload };
}

/**
 * Handles the gnosis wildcard Pattern workflow.
 */
export function wildcardPattern(): WildcardPattern {
  return { type: 'wildcard' };
}

/**
 * Handles the gnosis literal Pattern workflow.
 */
export function literalPattern(value: unknown): LiteralPattern {
  return { type: 'literal', value };
}

// ============================================================================
// Binding extraction -- resolve a pattern against a value
// ============================================================================

export interface BindingResult {
  /** Whether the pattern matched */
  matched: boolean;
  /** Extracted bindings (name → value) */
  bindings: Record<string, unknown>;
  /** Diagnostic if match failed */
  diagnostic?: string;
}

function successBinding(bindings: Record<string, unknown>): BindingResult {
  return { matched: true, bindings };
}

function failBinding(diagnostic: string): BindingResult {
  return { matched: false, bindings: {}, diagnostic };
}

/**
 * Extract bindings from a value using a destructure pattern.
 */
export function extractBindings(
  pattern: DestructurePattern,
  value: unknown
): BindingResult {
  try {
    switch (pattern.type) {
      case 'wildcard':
        return successBinding({});

      case 'literal':
        return value === pattern.value
          ? successBinding({})
          : failBinding(
              `Expected literal ${String(pattern.value)}, got ${String(value)}`
            );

      case 'record':
        return extractRecordBindings(pattern, value);

      case 'tuple':
        return extractTupleBindings(pattern, value);

      case 'variant':
        return extractVariantBindings(pattern, value);

      case 'nested':
        return extractNestedBindings(pattern, value);
    }

  } catch (error) {
    throw new Error(`[extractBindings] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function extractRecordBindings(
  pattern: RecordPattern,
  value: unknown
): BindingResult {
  try {
    if (typeof value !== 'object' || value === null) {
      return failBinding(`Expected object, got ${typeof value}`);
    }

    const obj = value as Record<string, unknown>;
    const bindings: Record<string, unknown> = {};
    const boundKeys = new Set<string>();

    for (const [fieldName, binding] of Object.entries(pattern.fields)) {
      if (!(fieldName in obj)) {
        return failBinding(`Missing field '${fieldName}'`);
      }

      boundKeys.add(fieldName);

      if (typeof binding === 'string') {
        bindings[binding] = obj[fieldName];
      } else {
        // Nested pattern
        const nested = extractBindings(binding, obj[fieldName]);
        if (!nested.matched) return nested;
        Object.assign(bindings, nested.bindings);
      }
    }

    if (pattern.rest) {
      const rest: Record<string, unknown> = {};
      for (const key of Object.keys(obj)) {
        if (!boundKeys.has(key)) rest[key] = obj[key];
      }
      bindings[pattern.rest] = rest;
    }

    return successBinding(bindings);

  } catch (error) {
    throw new Error(`[extractRecordBindings] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function extractTupleBindings(
  pattern: TuplePattern,
  value: unknown
): BindingResult {
  try {
    if (!Array.isArray(value)) {
      return failBinding(`Expected array, got ${typeof value}`);
    }

    if (value.length < pattern.elements.length) {
      return failBinding(
        `Expected at least ${pattern.elements.length} elements, got ${value.length}`
      );
    }

    const bindings: Record<string, unknown> = {};

    for (let i = 0; i < pattern.elements.length; i++) {
      const element = pattern.elements[i];
      if (typeof element === 'string') {
        bindings[element] = value[i];
      } else {
        const nested = extractBindings(element, value[i]);
        if (!nested.matched) return nested;
        Object.assign(bindings, nested.bindings);
      }
    }

    return successBinding(bindings);

  } catch (error) {
    throw new Error(`[extractTupleBindings] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function extractVariantBindings(
  pattern: VariantPattern,
  value: unknown
): BindingResult {
  try {
    if (typeof value !== 'object' || value === null) {
      return failBinding(`Expected variant object, got ${typeof value}`);
    }

    const v = value as Variant;
    if (v.tag !== pattern.tag) {
      return failBinding(`Expected variant '${pattern.tag}', got '${v.tag}'`);
    }

    if (typeof pattern.payload === 'string') {
      return successBinding({ [pattern.payload]: v.value });
    }

    return extractBindings(pattern.payload, v.value);

  } catch (error) {
    throw new Error(`[extractVariantBindings] ${error instanceof Error ? error.message : String(error)}`);
  }
}

function extractNestedBindings(
  pattern: NestedPattern,
  value: unknown
): BindingResult {
  try {
    const outer = extractBindings(pattern.outer, value);
    if (!outer.matched) return outer;
    // Apply inner pattern to all bound values combined
    const inner = extractBindings(pattern.inner, outer.bindings);
    if (!inner.matched) return inner;
    return successBinding({ ...outer.bindings, ...inner.bindings });

  } catch (error) {
    throw new Error(`[extractNestedBindings] ${error instanceof Error ? error.message : String(error)}`);
  }
}

// ============================================================================
// Pattern validation -- check patterns against type definitions
// ============================================================================

export interface PatternValidation {
  valid: boolean;
  diagnostics: string[];
}

/**
 * Validate a variant pattern against a sum type definition.
 */
export function validateVariantPattern(
  pattern: VariantPattern,
  typeDef: SumTypeDefinition
): PatternValidation {
  try {
    const diagnostics: string[] = [];
    if (!typeDef.cases.includes(pattern.tag)) {
      diagnostics.push(
        `Tag '${pattern.tag}' is not a valid case of ${
          typeDef.name
        }. Valid cases: [${typeDef.cases.join(', ')}]`
      );
    }
    return { valid: diagnostics.length === 0, diagnostics };

  } catch (error) {
    throw new Error(`[validateVariantPattern] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Validate a record pattern against a product type definition.
 */
export function validateRecordPattern(
  pattern: RecordPattern,
  typeDef: ProductTypeDefinition
): PatternValidation {
  try {
    const diagnostics: string[] = [];
    const fieldNames = new Set(typeDef.fields.map((f) => f.name));

    for (const key of Object.keys(pattern.fields)) {
      if (!fieldNames.has(key)) {
        diagnostics.push(`Field '${key}' not found in ${typeDef.name}`);
      }
    }

    // Check required fields are present (unless rest is used)
    if (!pattern.rest) {
      for (const field of typeDef.fields) {
        if (field.required && !(field.name in pattern.fields)) {
          diagnostics.push(`Required field '${field.name}' not bound in pattern`);
        }
      }
    }

    return { valid: diagnostics.length === 0, diagnostics };

  } catch (error) {
    throw new Error(`[validateRecordPattern] ${error instanceof Error ? error.message : String(error)}`);
  }
}

/**
 * Check exhaustiveness of a set of variant patterns over a sum type.
 */
export function checkPatternExhaustiveness(
  patterns: VariantPattern[],
  typeDef: SumTypeDefinition
): { exhaustive: boolean; missing: string[] } {
  const covered = new Set(patterns.map((p) => p.tag));
  const missing = typeDef.cases.filter((c) => !covered.has(c));
  return { exhaustive: missing.length === 0, missing };
}

// ============================================================================
// Match expression -- pattern matching with destructuring
// ============================================================================

export interface MatchArmWithPattern<R> {
  pattern: DestructurePattern;
  guard?: (bindings: Record<string, unknown>) => boolean;
  handler: (bindings: Record<string, unknown>) => R;
}

/**
 * Match a value against a sequence of patterned arms.
 * First matching arm wins.
 */
export function matchPatterns<R>(
  value: unknown,
  arms: MatchArmWithPattern<R>[]
): { matched: boolean; result?: R; armIndex?: number } {
  try {
    for (let i = 0; i < arms.length; i++) {
      const arm = arms[i];
      const binding = extractBindings(arm.pattern, value);
      if (binding.matched) {
        if (arm.guard && !arm.guard(binding.bindings)) continue;
        return {
          matched: true,
          result: arm.handler(binding.bindings),
          armIndex: i,
        };
      }
    }
    return { matched: false };

  } catch (error) {
    throw new Error(`[matchPatterns] ${error instanceof Error ? error.message : String(error)}`);
  }
}
