// @a0n/.ignored_gnosis/src/differentiable-handlers.ts
/**
 * Differentiable Programming GG Topology Handlers
 *
 * Handlers for gradient tape operations, parameter creation,
 * differentiable ops, loss computation, and gradient analysis.
 */

import type { GnosisRegistry } from './runtime/registry';
import {
  GradientTape,
  parameter,
  constant,
  diffValue,
  add,
  mul,
  sub,
  neg,
  relu,
  sigmoid,
  tanh,
  exp,
  log,
  pow,
  mseLoss,
  bceLoss,
  createOptimizer,
  optimizerStep,
  analyzeGradientFlow,
  type DiffValue,
  type OptimizerConfig,
  type OptimizerState,
} from './differentiable';

/**
 * Handles the gnosis register Differentiable Handlers workflow.
 */
export function registerDifferentiableHandlers(registry: GnosisRegistry): void {
  // Shared tape and optimizer state per execution
  let activeTape = new GradientTape();
  let activeOptimizer: OptimizerState | null = null;
  const paramIndices: number[] = [];

  // Initialize a fresh gradient tape
  registry.register('GradientTapeInit', async (_payload, props) => {
    activeTape = new GradientTape();
    paramIndices.length = 0;
    const lr = parseFloat(props.learning_rate ?? '0.01');
    const momentum = parseFloat(props.momentum ?? '0');
    const decay = parseFloat(props.weight_decay ?? '0');
    activeOptimizer = createOptimizer({
      learningRate: lr,
      momentum,
      weightDecay: decay,
    });
    return {
      initialized: true,
      learningRate: lr,
      momentum,
      weightDecay: decay,
    };
  });

  // Create a parameter (tracked, requires grad)
  registry.register('Parameter', async (payload, props) => {
    const value = parseFloat(props.value ?? String(payload ?? '0'));
    const label = props.label ?? undefined;
    const p = parameter(value, label);
    const idx = activeTape.track(p);
    paramIndices.push(idx);
    return { index: idx, value: p.data, label, requiresGrad: true };
  });

  // Create a constant (not tracked)
  registry.register('Constant', async (payload, props) => {
    const value = parseFloat(props.value ?? String(payload ?? '0'));
    const label = props.label ?? undefined;
    const c = constant(value, label);
    const idx = activeTape.track(c);
    return { index: idx, value: c.data, label, requiresGrad: false };
  });

  // Differentiable operations
  registry.register('DiffAdd', async (payload, props) => {
    const { a, b } = payload as { a: number; b: number };
    const result = add(activeTape, a, b);
    return { index: result, value: activeTape.get(result).data };
  });

  registry.register('DiffMul', async (payload, props) => {
    const { a, b } = payload as { a: number; b: number };
    const result = mul(activeTape, a, b);
    return { index: result, value: activeTape.get(result).data };
  });

  registry.register('DiffSub', async (payload, props) => {
    const { a, b } = payload as { a: number; b: number };
    const result = sub(activeTape, a, b);
    return { index: result, value: activeTape.get(result).data };
  });

  registry.register('DiffReLU', async (payload) => {
    const idx = (payload as { index: number }).index;
    const result = relu(activeTape, idx);
    return { index: result, value: activeTape.get(result).data };
  });

  registry.register('DiffSigmoid', async (payload) => {
    const idx = (payload as { index: number }).index;
    const result = sigmoid(activeTape, idx);
    return { index: result, value: activeTape.get(result).data };
  });

  registry.register('DiffTanh', async (payload) => {
    const idx = (payload as { index: number }).index;
    const result = tanh(activeTape, idx);
    return { index: result, value: activeTape.get(result).data };
  });

  // Loss functions
  registry.register('MSELoss', async (payload) => {
    const { prediction, target } = payload as {
      prediction: number;
      target: number;
    };
    const result = mseLoss(activeTape, prediction, target);
    return { index: result, value: activeTape.get(result).data };
  });

  // Backward pass
  registry.register('Backward', async (payload) => {
    const lossIdx = (payload as { index: number }).index;
    activeTape.backward(lossIdx);
    const params = activeTape.parameters();
    return {
      gradients: params.map((p) => ({
        value: p.data,
        grad: p.grad,
        label: p.label,
      })),
    };
  });

  // Optimizer step
  registry.register('OptimizerStep', async () => {
    if (!activeOptimizer) {
      return { error: 'No optimizer initialized' };
    }
    optimizerStep(activeTape, activeOptimizer, paramIndices);
    const params = paramIndices.map((i) => {
      const p = activeTape.get(i);
      return { index: i, value: p.data, grad: p.grad, label: p.label };
    });
    return { step: activeOptimizer.step, params };
  });

  // Zero gradients
  registry.register('ZeroGrad', async () => {
    activeTape.zeroGrad();
    return { zeroed: true };
  });

  // Gradient flow analysis
  registry.register('GradientAnalysis', async () => {
    return analyzeGradientFlow(activeTape);
  });

  // Full training step: forward → loss → backward → step
  registry.register('TrainStep', async (payload, props) => {
    const { forwardFn, targetIdx } = payload as {
      forwardFn: (tape: GradientTape, paramIdxs: number[]) => number;
      targetIdx: number;
    };

    activeTape.zeroGrad();
    const predIdx = forwardFn(activeTape, paramIndices);
    const lossIdx = mseLoss(activeTape, predIdx, targetIdx);
    activeTape.backward(lossIdx);

    if (activeOptimizer) {
      optimizerStep(activeTape, activeOptimizer, paramIndices);
    }

    return {
      loss: activeTape.get(lossIdx).data,
      step: activeOptimizer?.step ?? 0,
      flow: analyzeGradientFlow(activeTape),
    };
  });
}

export const DIFFERENTIABLE_OPS = [
  'add',
  'mul',
  'sub',
  'neg',
  'relu',
  'sigmoid',
  'tanh',
  'exp',
  'log',
  'pow',
  'mseLoss',
  'bceLoss',
] as const;
