// @a0n/.ignored_gnosis/src/dockerfile-topology.ts
/**
 * dockerfile-topology.ts -- Parse Dockerfiles into GG topologies
 *
 * A Dockerfile is already a linear topology. Multi-stage Dockerfiles are DAGs.
 * This module translates Dockerfile instructions into GG topology source,
 * mapping Docker semantics onto fork/race/fold/vent primitives:
 *
 *   FROM image AS stage  →  fork node (new stage subgraph)
 *   RUN command          →  PROCESS edge (sequential fold within stage)
 *   COPY --from=stage    →  RACE edge (cross-branch artifact selection)
 *   EXPOSE port          →  VENT node (declared dissipation)
 *   CMD / ENTRYPOINT     →  terminal sink node
 *   ENV / ARG / WORKDIR  →  property annotations on enclosing stage
 *
 * Pure parser -- no external deps, no execution, no Docker daemon.
 *
 * @module @a0n/gnosis/dockerfile-topology
 */

// ============================================================================
// Types
// ============================================================================

export type DockerfileInstructionType =
  | 'FROM'
  | 'RUN'
  | 'COPY'
  | 'ADD'
  | 'EXPOSE'
  | 'ENV'
  | 'ARG'
  | 'WORKDIR'
  | 'CMD'
  | 'ENTRYPOINT'
  | 'LABEL'
  | 'VOLUME'
  | 'USER'
  | 'HEALTHCHECK'
  | 'SHELL'
  | 'STOPSIGNAL'
  | 'ONBUILD';

export interface DockerfileInstruction {
  readonly type: DockerfileInstructionType;
  readonly args: string;
  readonly lineNumber: number;
  /** Stage name if this instruction belongs to a named stage */
  readonly stage: string;
  /** For COPY --from=stage, the source stage name */
  readonly fromStage?: string;
}

export interface DockerfileStage {
  readonly name: string;
  readonly image: string;
  readonly instructions: readonly DockerfileInstruction[];
  readonly lineNumber: number;
}

export interface DockerfileTopology {
  /** Generated GG topology source */
  readonly ggSource: string;
  /** Parsed instructions */
  readonly instructions: readonly DockerfileInstruction[];
  /** Multi-stage build stage names */
  readonly stages: readonly DockerfileStage[];
  /** Exposed ports */
  readonly exposedPorts: readonly number[];
  /** Whether this is a multi-stage build */
  readonly isMultiStage: boolean;
  /** Capability requirements inferred from instructions */
  readonly requiredCapabilities: readonly string[];
}

// ============================================================================
// Parser
// ============================================================================

/**
 * Parse a Dockerfile into structured instructions.
 *
 * Handles continuation lines (backslash-newline), comments, and all
 * standard Dockerfile instructions.
 */
export function parseDockerfile(
  source: string
): readonly DockerfileInstruction[] {
  const lines = source.split('\n');
  const instructions: DockerfileInstruction[] = [];
  let currentStage = 'default';
  let accumulated = '';
  let startLine = 0;

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    const trimmed = line.trim();

    // Skip empty lines and comments (but don't break continuations)
    if (trimmed === '' || trimmed.startsWith('#')) {
      // Docker spec: blank lines within continuations are allowed
      if (accumulated && trimmed === '') continue;
      // Comments end continuations
      if (accumulated && trimmed.startsWith('#')) {
        const inst = parseInstruction(accumulated, startLine, currentStage);
        if (inst) {
          if (inst.type === 'FROM') {
            currentStage =
              extractStageName(inst.args) ??
              `stage_${instructions.filter((i) => i.type === 'FROM').length}`;
          }
          instructions.push({ ...inst, stage: currentStage });
        }
        accumulated = '';
      }
      continue;
    }

    // Handle continuation lines
    if (accumulated) {
      if (trimmed.endsWith('\\')) {
        accumulated += ' ' + trimmed.slice(0, -1).trim();
        continue;
      } else {
        accumulated += ' ' + trimmed;
        const inst = parseInstruction(accumulated, startLine, currentStage);
        if (inst) {
          if (inst.type === 'FROM') {
            currentStage =
              extractStageName(inst.args) ??
              `stage_${instructions.filter((i) => i.type === 'FROM').length}`;
          }
          instructions.push({ ...inst, stage: currentStage });
        }
        accumulated = '';
        continue;
      }
    }

    // Start of a new instruction
    if (trimmed.endsWith('\\')) {
      accumulated = trimmed.slice(0, -1).trim();
      startLine = i + 1;
      continue;
    }

    const inst = parseInstruction(trimmed, i + 1, currentStage);
    if (inst) {
      if (inst.type === 'FROM') {
        currentStage =
          extractStageName(inst.args) ??
          `stage_${instructions.filter((i) => i.type === 'FROM').length}`;
      }
      instructions.push({ ...inst, stage: currentStage });
    }
  }

  // Flush any remaining continuation
  if (accumulated) {
    const inst = parseInstruction(accumulated, startLine, currentStage);
    if (inst) {
      instructions.push({ ...inst, stage: currentStage });
    }
  }

  return instructions;
}

function parseInstruction(
  line: string,
  lineNumber: number,
  currentStage: string
): DockerfileInstruction | null {
  const spaceIdx = line.indexOf(' ');
  if (spaceIdx === -1) {
    // Instructions like FROM with no args (invalid but handle gracefully)
    const type = line.toUpperCase();
    if (isValidInstruction(type)) {
      return {
        type: type as DockerfileInstructionType,
        args: '',
        lineNumber,
        stage: currentStage,
      };
    }
    return null;
  }

  const type = line.slice(0, spaceIdx).toUpperCase();
  const args = line.slice(spaceIdx + 1).trim();

  if (!isValidInstruction(type)) return null;

  const inst: DockerfileInstruction = {
    type: type as DockerfileInstructionType,
    args,
    lineNumber,
    stage: currentStage,
  };

  // Extract --from for COPY/ADD
  if ((type === 'COPY' || type === 'ADD') && args.startsWith('--from=')) {
    const fromEnd = args.indexOf(' ');
    const fromStage = args.slice(7, fromEnd > 0 ? fromEnd : undefined).trim();
    if (fromStage) return { ...inst, fromStage };
  }

  return inst;
}

function isValidInstruction(type: string): boolean {
  return [
    'FROM',
    'RUN',
    'COPY',
    'ADD',
    'EXPOSE',
    'ENV',
    'ARG',
    'WORKDIR',
    'CMD',
    'ENTRYPOINT',
    'LABEL',
    'VOLUME',
    'USER',
    'HEALTHCHECK',
    'SHELL',
    'STOPSIGNAL',
    'ONBUILD',
  ].includes(type);
}

function extractStageName(fromArgs: string): string | null {
  const lower = fromArgs.toLowerCase();
  const asIdx = lower.indexOf(' as ');
  if (asIdx === -1) return null;
  return fromArgs.slice(asIdx + 4).trim();
}

// ============================================================================
// Stage Extraction
// ============================================================================

/**
 * Group instructions into stages.
 */
export function extractStages(
  instructions: readonly DockerfileInstruction[]
): readonly DockerfileStage[] {
  const stages: DockerfileStage[] = [];
  let current: {
    name: string;
    image: string;
    instructions: DockerfileInstruction[];
    lineNumber: number;
  } | null = null;

  for (const inst of instructions) {
    if (inst.type === 'FROM') {
      if (current) stages.push(current);
      const name = extractStageName(inst.args) ?? `stage_${stages.length}`;
      const image = inst.args.split(/\s+/)[0];
      current = {
        name,
        image,
        instructions: [inst],
        lineNumber: inst.lineNumber,
      };
    } else if (current) {
      current.instructions.push(inst);
    }
  }

  if (current) stages.push(current);
  return stages;
}

// ============================================================================
// GG Topology Generation
// ============================================================================

/**
 * Generate GG topology source from a Dockerfile.
 *
 * Maps Docker semantics onto fork/race/fold/vent primitives.
 */
export function generateGGTopology(
  instructions: readonly DockerfileInstruction[]
): string {
  const stages = extractStages(instructions);
  const lines: string[] = [];

  lines.push('// Generated from Dockerfile -- DO NOT EDIT');
  lines.push(
    '// Each stage is a fork branch. COPY --from is a cross-branch race.'
  );
  lines.push('');

  for (const stage of stages) {
    lines.push(`// Stage: ${stage.name} (${stage.image})`);
    lines.push(
      `(${sanitizeId(stage.name)} : Stage { image: '${escapeGG(
        stage.image
      )}' })`
    );

    let prevNode = sanitizeId(stage.name);
    let nodeIdx = 0;

    for (const inst of stage.instructions) {
      if (inst.type === 'FROM') continue; // Already handled as stage declaration

      const nodeId = `${sanitizeId(
        stage.name
      )}_${inst.type.toLowerCase()}_${nodeIdx++}`;

      switch (inst.type) {
        case 'RUN': {
          const cmd = escapeGG(inst.args);
          lines.push(
            `(${prevNode}) -[:PROCESS]-> (${nodeId} : LayerRun { cmd: '${cmd}' })`
          );
          prevNode = nodeId;
          break;
        }
        case 'COPY':
        case 'ADD': {
          if (inst.fromStage) {
            // Cross-stage COPY -- this is a RACE edge
            const fromNode = sanitizeId(inst.fromStage);
            const parts = inst.args
              .split(/\s+/)
              .filter((p) => !p.startsWith('--'));
            const src = parts[0] ?? '.';
            const dest = parts[1] ?? '.';
            lines.push(
              `(${fromNode}, ${prevNode}) -[:RACE]-> (${nodeId} : LayerCopy { from: '${escapeGG(
                inst.fromStage
              )}', src: '${escapeGG(src)}', dest: '${escapeGG(dest)}' })`
            );
          } else {
            const parts = inst.args
              .split(/\s+/)
              .filter((p) => !p.startsWith('--'));
            const src = parts[0] ?? '.';
            const dest = parts[1] ?? '.';
            lines.push(
              `(${prevNode}) -[:PROCESS]-> (${nodeId} : Layer${
                inst.type === 'COPY' ? 'Copy' : 'Add'
              } { src: '${escapeGG(src)}', dest: '${escapeGG(dest)}' })`
            );
          }
          prevNode = nodeId;
          break;
        }
        case 'EXPOSE': {
          const port = inst.args.trim();
          lines.push(
            `(${prevNode}) -[:VENT { port: ${port} }]-> (${nodeId} : Exposed)`
          );
          prevNode = nodeId;
          break;
        }
        case 'CMD':
        case 'ENTRYPOINT': {
          const cmd = escapeGG(inst.args);
          lines.push(
            `(${prevNode}) -[:PROCESS]-> (${nodeId} : ${
              inst.type === 'CMD' ? 'Cmd' : 'Entrypoint'
            } { cmd: '${cmd}' })`
          );
          prevNode = nodeId;
          break;
        }
        case 'ENV':
        case 'ARG':
        case 'WORKDIR':
        case 'USER':
        case 'LABEL':
        case 'VOLUME': {
          // Property annotations -- attach to previous node
          lines.push(
            `(${prevNode}) { ${inst.type.toLowerCase()}: '${escapeGG(
              inst.args
            )}' }`
          );
          break;
        }
        default:
          // HEALTHCHECK, SHELL, STOPSIGNAL, ONBUILD -- property annotations
          lines.push(
            `(${prevNode}) { ${inst.type.toLowerCase()}: '${escapeGG(
              inst.args
            )}' }`
          );
          break;
      }
    }

    lines.push('');
  }

  return lines.join('\n');
}

function sanitizeId(name: string): string {
  return name.replace(/[^a-zA-Z0-9_]/g, '_');
}

function escapeGG(value: string): string {
  return value.replace(/'/g, "\\'").replace(/\n/g, '\\n');
}

// ============================================================================
// Full Pipeline
// ============================================================================

/**
 * Parse a Dockerfile and generate its GG topology.
 */
export function dockerfileToTopology(source: string): DockerfileTopology {
  const instructions = parseDockerfile(source);
  const stages = extractStages(instructions);
  const ggSource = generateGGTopology(instructions);

  const exposedPorts: number[] = [];
  for (const inst of instructions) {
    if (inst.type === 'EXPOSE') {
      const port = parseInt(inst.args.trim(), 10);
      if (!isNaN(port)) exposedPorts.push(port);
    }
  }

  // Infer capability requirements
  const requiredCapabilities: string[] = ['process.spawn'];
  const hasFromStage = instructions.some((i) => i.fromStage);
  if (hasFromStage) requiredCapabilities.push('process.docker');

  return {
    ggSource,
    instructions,
    stages,
    exposedPorts,
    isMultiStage: stages.length > 1,
    requiredCapabilities,
  };
}
