// @a0n/.ignored_gnosis/src/earth-chronometry.ts
/**
 * Earth formation age in **Gnostic time coordinates** (picolorenzo ladder).
 *
 * ## What is formalized here
 *
 * - The **conversion** from Julian years → seconds → picolorenzos, using the same
 *   contract as the rest of the Gnostic time system (`1 pLo = π` days).
 *   Same identity is proved in Lean as `julianYears_eq_div_picolorenzo` in
 *   `open-source/gnosis/lean/Lean/PlanetaryHomologySandbox/PicolorenzoJulianSketch.lean`
 *   (`Real.pi × 86400` seconds per picolorenzo).
 * - A **structured report**: duration in pLo, nanolorenzos (1000 pLo), and the
 *   `decomposeGnosticTime` ladder (Gnostic “aeon” steps = 110 pLo each — not ICS
 *   geologic eons).
 *
 * ## What is *not* formalized
 *
 * The numeric **age of Earth** (~4.54 billion years) is **not** a consequence of
 * Gnostic numerology. It is the current **geochronology consensus** from
 * concordant radiometric dating of the oldest solid materials (CAIs in meteorites,
 * ancient terrestrial zircons). This module takes that duration as an explicit
 * **input constant** and only re-expresses it.
 *
 * @see Dalrymple (2004) for classical synthesis; planetary science textbooks cite
 * ~4.54 Ga with small quoted uncertainties (order ±1–2%).
 */

import {
  decomposeGnosticTime,
  formatGnosticTime,
  formatPicolorenzos,
  secondsToPicolorenzos,
} from './math';

/** Julian year in SI seconds (no leap-second table). */
export const JULIAN_YEAR_SECONDS = 365.25 * 86400;

/**
 * Convert a span measured in Julian years to picolorenzos.
 * `1 pLo = π days`, so `years × 365.25 d / (π d/pLo) = years × 365.25 / π` pLo.
 */
export function julianYearsToPicolorenzos(years: number): number {
  return secondsToPicolorenzos(years * JULIAN_YEAR_SECONDS);
}

/**
 * Canonical **mid-range** formation age used for pedagogy (billion Julian years).
 * Not a Gnosis inference: geochronology input.
 */
export const EARTH_FORMATION_AGE_GYR = 4.54;

/**
 * Same age in Julian years (scalar).
 */
export const EARTH_FORMATION_AGE_YEARS = EARTH_FORMATION_AGE_GYR * 1e9;

/**
 * Optional literature-style uncertainty on the **input** age (±50 Myr on 4.54 Ga),
 * expressed only for bracketing — still not computed from axioms.
 */
export const EARTH_FORMATION_AGE_YEARS_UNCERTAINTY_ONE_SIGMA = 0.05e9;

/** One nanolorenzo spans 1000 picolorenzos of duration (see `NANOLORENZO_SECONDS`). */
export const NANOLORENZOS_PER_PICOLORENZO = 1000;

export interface EarthFormationGnosticReport {
  readonly epistemicNote: 'geochronology_input_not_gnosis_theorem';
  readonly formationAgeJulianYears: number;
  readonly formationAgeGyr: number;
  /** Duration in picolorenzos (π-day units). */
  readonly durationPicolorenzos: number;
  /** Duration in nanolorenzos (1 nLo = 1000 pLo). */
  readonly durationNanolorenzos: number;
  /**
   * Gnostic ladder decomposition of `durationPicolorenzos`.
   * Field `aeons` counts **110 pLo** blocks (see `2 * PLEROMA`), not Phanerozoic-scale geologic eons.
   */
  readonly gnosticLadder: ReturnType<typeof decomposeGnosticTime>;
  readonly formattedGnosticLadder: string;
  readonly formattedPicolorenzos: string;
}

/**
 * Full report mapping the consensus formation age into Gnostic duration coordinates.
 */
export function earthFormationAgeGnosticReport(
  formationAgeJulianYears: number = EARTH_FORMATION_AGE_YEARS
): EarthFormationGnosticReport {
  const durationPicolorenzos = julianYearsToPicolorenzos(formationAgeJulianYears);
  const durationNanolorenzos = durationPicolorenzos / NANOLORENZOS_PER_PICOLORENZO;
  const gnosticLadder = decomposeGnosticTime(durationPicolorenzos);

  return {
    epistemicNote: 'geochronology_input_not_gnosis_theorem',
    formationAgeJulianYears,
    formationAgeGyr: formationAgeJulianYears / 1e9,
    durationPicolorenzos,
    durationNanolorenzos,
    gnosticLadder,
    formattedGnosticLadder: formatGnosticTime(durationPicolorenzos),
    formattedPicolorenzos: formatPicolorenzos(durationPicolorenzos),
  };
}

/**
 * Bracket the consensus age by ±`EARTH_FORMATION_AGE_YEARS_UNCERTAINTY_ONE_SIGMA` (input uncertainty only).
 */
export function earthFormationAgeUncertaintyBracketGnostic(): {
  readonly lowerYears: number;
  readonly upperYears: number;
  readonly lowerPLo: number;
  readonly upperPLo: number;
} {
  const d = EARTH_FORMATION_AGE_YEARS_UNCERTAINTY_ONE_SIGMA;
  const mid = EARTH_FORMATION_AGE_YEARS;
  const lowerYears = mid - d;
  const upperYears = mid + d;
  return {
    lowerYears,
    upperYears,
    lowerPLo: julianYearsToPicolorenzos(lowerYears),
    upperPLo: julianYearsToPicolorenzos(upperYears),
  };
}
