// @a0n/.ignored_gnosis/src/effects.ts
/**
 * effects.ts -- Explicit effect system for gnosis topologies
 *
 * Effects make host requirements visible in topology signatures and
 * module exports. Declared capabilities and inferred capabilities
 * unify into one effect model. Target validation becomes a type check.
 *
 * Roadmap item #1: the dependency gate for structured concurrency,
 * quantum features, and differentiable programming.
 */

// ============================================================================
// Effect kinds -- the concrete operational effects
// ============================================================================

/**
 * All known effect kinds in gnosis. Each maps to a host capability
 * that must be present at the deployment target.
 */
export type EffectKind =
  | 'fs.local'
  | 'fs.durable'
  | 'net.tcp.client'
  | 'net.tcp.server'
  | 'net.udp'
  | 'auth.ucan'
  | 'auth.zk'
  | 'auth.custodial'
  | 'render.dom'
  | 'render.canvas'
  | 'render.webgpu'
  | 'neural.cpu'
  | 'neural.gpu'
  | 'neural.webnn'
  | 'quantum.sim'
  | 'quantum.hardware'
  | 'diff.forward'
  | 'diff.backward'
  | 'time.timeout'
  | 'time.deadline'
  | 'time.interval'
  | 'crypto.random'
  | 'crypto.sign'
  | 'crypto.verify';

/** All valid effect kinds */
export const ALL_EFFECT_KINDS: readonly EffectKind[] = [
  'fs.local',
  'fs.durable',
  'net.tcp.client',
  'net.tcp.server',
  'net.udp',
  'auth.ucan',
  'auth.zk',
  'auth.custodial',
  'render.dom',
  'render.canvas',
  'render.webgpu',
  'neural.cpu',
  'neural.gpu',
  'neural.webnn',
  'quantum.sim',
  'quantum.hardware',
  'diff.forward',
  'diff.backward',
  'time.timeout',
  'time.deadline',
  'time.interval',
  'crypto.random',
  'crypto.sign',
  'crypto.verify',
];

/**
 * Returns whether is Effect Kind is true.
 */
export function isEffectKind(s: string): s is EffectKind {
  return (ALL_EFFECT_KINDS as readonly string[]).includes(s);
}

// ============================================================================
// Effect signature -- what a node or topology requires
// ============================================================================

export interface EffectRequirement {
  kind: EffectKind;
  source: 'declared' | 'inferred';
  nodeId: string;
  reason: string;
}

export interface EffectSignature {
  /** All effect requirements (declared + inferred) */
  requirements: EffectRequirement[];
  /** Unique effect kinds */
  kinds: EffectKind[];
  /** Whether the signature is pure (no effects) */
  pure: boolean;
}

/**
 * Creates the Effect Signature.
 */
export function createEffectSignature(
  requirements: EffectRequirement[]
): EffectSignature {
  const kinds = [...new Set(requirements.map((r) => r.kind))];
  return {
    requirements,
    kinds,
    pure: kinds.length === 0,
  };
}

// ============================================================================
// Effect inference -- derive effects from node labels and properties
// ============================================================================

const LABEL_EFFECT_MAP: Record<string, EffectKind[]> = {
  FileReader: ['fs.local'],
  FileWriter: ['fs.local'],
  DurableStore: ['fs.durable'],
  DurableObject: ['fs.durable'],
  KVStore: ['fs.durable'],
  HttpClient: ['net.tcp.client'],
  HttpServer: ['net.tcp.server'],
  WebSocket: ['net.tcp.client'],
  UdpSocket: ['net.udp'],
  UCANVerifier: ['auth.ucan'],
  UCANIssuer: ['auth.ucan'],
  ZKProver: ['auth.zk'],
  ZKVerifier: ['auth.zk'],
  Renderer: ['render.dom'],
  CanvasRenderer: ['render.canvas'],
  WebGPURenderer: ['render.webgpu'],
  NeuralEngine: ['neural.cpu'],
  GPUEngine: ['neural.gpu'],
  WebNNEngine: ['neural.webnn'],
  QuantumCircuit: ['quantum.sim'],
  QuantumDevice: ['quantum.hardware'],
  DifferentiableNode: ['diff.forward', 'diff.backward'],
  GradientNode: ['diff.backward'],
  Timer: ['time.timeout'],
  Scheduler: ['time.interval'],
  Signer: ['crypto.sign'],
  Verifier: ['crypto.verify'],
};

const PROPERTY_EFFECT_MAP: Record<string, EffectKind[]> = {
  timeout: ['time.timeout'],
  deadline: ['time.deadline'],
  interval: ['time.interval'],
  transport: ['net.tcp.client'],
  protocol: ['net.tcp.client'],
  render_target: ['render.dom'],
  gpu_backend: ['neural.gpu'],
  learning_rate: ['diff.forward', 'diff.backward'],
  gradient: ['diff.backward'],
};

export interface NodeDescriptor {
  id: string;
  labels: string[];
  properties: Record<string, string>;
}

/**
 * Infer effects from a node's labels and properties.
 */
export function inferNodeEffects(node: NodeDescriptor): EffectRequirement[] {
  const effects: EffectRequirement[] = [];
  const seen = new Set<EffectKind>();

  // Infer from labels
  for (const label of node.labels) {
    const mapped = LABEL_EFFECT_MAP[label];
    if (mapped) {
      for (const kind of mapped) {
        if (!seen.has(kind)) {
          seen.add(kind);
          effects.push({
            kind,
            source: 'inferred',
            nodeId: node.id,
            reason: `label '${label}' implies ${kind}`,
          });
        }
      }
    }
  }

  // Infer from properties
  for (const [prop, value] of Object.entries(node.properties)) {
    const mapped = PROPERTY_EFFECT_MAP[prop];
    if (mapped) {
      for (const kind of mapped) {
        if (!seen.has(kind)) {
          seen.add(kind);
          effects.push({
            kind,
            source: 'inferred',
            nodeId: node.id,
            reason: `property '${prop}=${value}' implies ${kind}`,
          });
        }
      }
    }
  }

  return effects;
}

/**
 * Parse declared effects from a node's 'effects' or 'effect' property.
 */
export function parseDeclaredEffects(
  node: NodeDescriptor
): EffectRequirement[] {
  const raw = node.properties['effects'] ?? node.properties['effect'] ?? '';
  if (!raw) return [];

  return raw
    .split(/[,|]/)
    .map((s) => s.trim())
    .filter(isEffectKind)
    .map((kind) => ({
      kind,
      source: 'declared' as const,
      nodeId: node.id,
      reason: `declared in effects property`,
    }));
}

// ============================================================================
// Effect validation -- check declared vs inferred, target compatibility
// ============================================================================

export type RuntimeTarget =
  | 'agnostic'
  | 'workers'
  | 'tauri-edge'
  | 'node'
  | 'bun'
  | 'gnode'
  | 'wasm-browser'
  | 'wasi';

/** Effects supported per target */
const TARGET_SUPPORT: Record<RuntimeTarget, Set<EffectKind>> = {
  agnostic: new Set([
    'time.timeout',
    'time.deadline',
    'time.interval',
    'crypto.random',
    'crypto.sign',
    'crypto.verify',
  ]),
  workers: new Set([
    'fs.durable',
    'net.tcp.client',
    'auth.ucan',
    'auth.zk',
    'render.dom',
    'render.canvas',
    'neural.cpu',
    'neural.webnn',
    'quantum.sim',
    'diff.forward',
    'diff.backward',
    'time.timeout',
    'time.deadline',
    'time.interval',
    'crypto.random',
    'crypto.sign',
    'crypto.verify',
  ]),
  'tauri-edge': new Set([
    'fs.local',
    'fs.durable',
    'auth.ucan',
    'auth.zk',
    'render.dom',
    'render.canvas',
    'neural.cpu',
    'neural.webnn',
    'quantum.sim',
    'diff.forward',
    'diff.backward',
    'time.timeout',
    'time.deadline',
    'time.interval',
    'crypto.random',
    'crypto.sign',
    'crypto.verify',
  ]),
  node: new Set([
    'fs.local',
    'fs.durable',
    'net.tcp.client',
    'net.tcp.server',
    'net.udp',
    'auth.ucan',
    'auth.zk',
    'auth.custodial',
    'neural.cpu',
    'neural.gpu',
    'quantum.sim',
    'diff.forward',
    'diff.backward',
    'time.timeout',
    'time.deadline',
    'time.interval',
    'crypto.random',
    'crypto.sign',
    'crypto.verify',
  ]),
  bun: new Set([
    'fs.local',
    'fs.durable',
    'net.tcp.client',
    'net.tcp.server',
    'net.udp',
    'auth.ucan',
    'auth.zk',
    'auth.custodial',
    'neural.cpu',
    'neural.gpu',
    'quantum.sim',
    'diff.forward',
    'diff.backward',
    'time.timeout',
    'time.deadline',
    'time.interval',
    'crypto.random',
    'crypto.sign',
    'crypto.verify',
  ]),
  gnode: new Set([
    'fs.local',
    'fs.durable',
    'net.tcp.client',
    'net.tcp.server',
    'net.udp',
    'auth.ucan',
    'auth.zk',
    'auth.custodial',
    'neural.cpu',
    'neural.gpu',
    'quantum.sim',
    'diff.forward',
    'diff.backward',
    'time.timeout',
    'time.deadline',
    'time.interval',
    'crypto.random',
    'crypto.sign',
    'crypto.verify',
  ]),
  'wasm-browser': new Set([
    'auth.ucan',
    'auth.zk',
    'render.dom',
    'render.canvas',
    'neural.cpu',
    'neural.webnn',
    'quantum.sim',
    'diff.forward',
    'diff.backward',
    'time.timeout',
    'time.deadline',
    'time.interval',
    'crypto.random',
    'crypto.sign',
    'crypto.verify',
  ]),
  wasi: new Set([
    'fs.local',
    'auth.ucan',
    'auth.zk',
    'neural.cpu',
    'quantum.sim',
    'diff.forward',
    'diff.backward',
    'time.timeout',
    'time.deadline',
    'time.interval',
    'crypto.random',
    'crypto.sign',
    'crypto.verify',
  ]),
};

export interface EffectDiagnostic {
  severity: 'error' | 'warning';
  nodeId: string;
  message: string;
  kind?: EffectKind;
}

export interface EffectValidationResult {
  ok: boolean;
  signature: EffectSignature;
  diagnostics: EffectDiagnostic[];
  /** Effects inferred but not declared (incomplete contract) */
  undeclared: EffectRequirement[];
  /** Effects declared but not inferred (unused declarations) */
  unused: EffectRequirement[];
  /** Effects incompatible with target */
  unsupported: EffectRequirement[];
}

/**
 * Validate effects for a topology given a set of nodes and a target.
 */
export function validateEffects(
  nodes: NodeDescriptor[],
  target: RuntimeTarget = 'agnostic'
): EffectValidationResult {
  const diagnostics: EffectDiagnostic[] = [];
  const allRequirements: EffectRequirement[] = [];

  for (const node of nodes) {
    const declared = parseDeclaredEffects(node);
    const inferred = inferNodeEffects(node);
    allRequirements.push(...declared, ...inferred);

    // Check: inferred effects not in declared set
    const declaredKinds = new Set(declared.map((d) => d.kind));
    const inferredKinds = new Set(inferred.map((i) => i.kind));

    for (const inf of inferred) {
      if (!declaredKinds.has(inf.kind)) {
        diagnostics.push({
          severity: 'error',
          nodeId: node.id,
          message: `Effect '${inf.kind}' inferred (${inf.reason}) but not declared`,
          kind: inf.kind,
        });
      }
    }

    // Check: declared effects not inferred (unused)
    for (const dec of declared) {
      if (!inferredKinds.has(dec.kind)) {
        diagnostics.push({
          severity: 'warning',
          nodeId: node.id,
          message: `Effect '${dec.kind}' declared but not inferred from usage`,
          kind: dec.kind,
        });
      }
    }
  }

  // Check target compatibility
  const supported = TARGET_SUPPORT[target];
  const unsupported: EffectRequirement[] = [];
  for (const req of allRequirements) {
    if (!supported.has(req.kind)) {
      unsupported.push(req);
      diagnostics.push({
        severity: 'error',
        nodeId: req.nodeId,
        message: `Effect '${req.kind}' not supported on target '${target}'`,
        kind: req.kind,
      });
    }
  }

  const signature = createEffectSignature(allRequirements);
  const undeclared = allRequirements.filter(
    (r) =>
      r.source === 'inferred' &&
      !allRequirements.some(
        (d) =>
          d.source === 'declared' && d.kind === r.kind && d.nodeId === r.nodeId
      )
  );
  const unused = allRequirements.filter(
    (r) =>
      r.source === 'declared' &&
      !allRequirements.some(
        (i) =>
          i.source === 'inferred' && i.kind === r.kind && i.nodeId === r.nodeId
      )
  );

  return {
    ok: diagnostics.filter((d) => d.severity === 'error').length === 0,
    signature,
    diagnostics,
    undeclared,
    unused,
    unsupported,
  };
}

// ============================================================================
// Effect contract -- module-level effect declaration
// ============================================================================

export interface EffectContract {
  /** Module name */
  module: string;
  /** All effects this module requires */
  requires: EffectKind[];
  /** All effects this module provides (handlers that satisfy effects) */
  provides: EffectKind[];
  /** Minimum target that supports all required effects */
  minimumTarget: RuntimeTarget;
}

/**
 * Build an effect contract from a topology's effect signature.
 */
export function buildEffectContract(
  module: string,
  signature: EffectSignature
): EffectContract {
  const requires = signature.kinds;

  // Determine minimum target
  let minimumTarget: RuntimeTarget = 'agnostic';
  for (const kind of requires) {
    if (!TARGET_SUPPORT['agnostic'].has(kind)) {
      if (TARGET_SUPPORT['workers'].has(kind)) {
        if (minimumTarget === 'agnostic') minimumTarget = 'workers';
      } else if (TARGET_SUPPORT['tauri-edge'].has(kind)) {
        if (minimumTarget === 'agnostic') minimumTarget = 'tauri-edge';
      } else if (TARGET_SUPPORT['wasm-browser'].has(kind)) {
        if (minimumTarget === 'agnostic') minimumTarget = 'wasm-browser';
      } else if (TARGET_SUPPORT['node'].has(kind)) {
        minimumTarget = 'node';
      } else if (TARGET_SUPPORT['wasi'].has(kind)) {
        minimumTarget = 'wasi';
      }
    }
  }

  return {
    module,
    requires,
    provides: [],
    minimumTarget,
  };
}

/**
 * Check if two effect contracts are compatible for composition.
 * The consumer's required effects must be a subset of the provider's
 * available effects (provided + target-native).
 */
export function contractsCompatible(
  consumer: EffectContract,
  provider: EffectContract,
  target: RuntimeTarget
): { compatible: boolean; missing: EffectKind[] } {
  const available = new Set([...provider.provides, ...TARGET_SUPPORT[target]]);
  const missing = consumer.requires.filter((k) => !available.has(k));
  return { compatible: missing.length === 0, missing };
}
