// @a0n/.ignored_gnosis/src/effects-handlers.ts
/**
 * Effect System GG Topology Handlers
 *
 * Handlers for effect inference, validation, and contract building
 * within GG topology execution.
 */

import type { GnosisRegistry } from './runtime/registry';
import {
  type NodeDescriptor,
  type EffectKind,
  type RuntimeTarget,
  inferNodeEffects,
  parseDeclaredEffects,
  validateEffects,
  createEffectSignature,
  buildEffectContract,
  contractsCompatible,
  isEffectKind,
  ALL_EFFECT_KINDS,
} from './effects';

/**
 * Handles the gnosis register Effect Handlers workflow.
 */
export function registerEffectHandlers(registry: GnosisRegistry): void {
  // Infer effects from a node descriptor
  registry.register('EffectInference', async (payload, props) => {
    const node: NodeDescriptor = payload ?? {
      id: props.nodeId ?? 'unknown',
      labels: (props.labels ?? '')
        .split(',')
        .map((s) => s.trim())
        .filter(Boolean),
      properties: { ...props },
    };
    const declared = parseDeclaredEffects(node);
    const inferred = inferNodeEffects(node);
    return {
      nodeId: node.id,
      declared: declared.map((r) => ({ kind: r.kind, reason: r.reason })),
      inferred: inferred.map((r) => ({ kind: r.kind, reason: r.reason })),
      all: [...new Set([...declared, ...inferred].map((r) => r.kind))],
    };
  });

  // Validate effects for a set of nodes against a target
  registry.register('EffectValidator', async (payload, props) => {
    const nodes = (payload as NodeDescriptor[]) ?? [];
    const target = (props.target ?? 'agnostic') as RuntimeTarget;
    return validateEffects(nodes, target);
  });

  // Build effect contract for a module
  registry.register('EffectContractBuilder', async (payload, props) => {
    const moduleName = props.module ?? 'unnamed';
    const nodes = (payload as NodeDescriptor[]) ?? [];
    const target = (props.target ?? 'agnostic') as RuntimeTarget;
    const validation = validateEffects(nodes, target);
    const contract = buildEffectContract(moduleName, validation.signature);
    return { contract, validation };
  });

  // Check compatibility between two effect contracts
  registry.register('EffectCompatibilityChecker', async (payload, props) => {
    const { consumer, provider } = payload as {
      consumer: ReturnType<typeof buildEffectContract>;
      provider: ReturnType<typeof buildEffectContract>;
    };
    const target = (props.target ?? 'agnostic') as RuntimeTarget;
    return contractsCompatible(consumer, provider, target);
  });

  // List all available effect kinds
  registry.register('EffectKindEnumerator', async () => {
    return { kinds: ALL_EFFECT_KINDS, count: ALL_EFFECT_KINDS.length };
  });
}

export const EFFECT_SYSTEM_FEATURES = [
  'effect-inference',
  'effect-validation',
  'effect-contracts',
  'target-compatibility',
  'module-effect-exports',
] as const;
