// @a0n/.ignored_gnosis/src/emergence.ts
/**
 * emergence.ts -- Phase transitions and emergent order in void walks
 *
 * The identification:
 *   - Phase transition = gait change (stand → trot → canter → gallop)
 *   - Order parameter = kurtosis of complement distribution
 *   - Critical point = kurtosis threshold where gait changes
 *   - Crystallization = complement distribution converges to delta
 *   - Symmetry breaking = one dimension dominates complement
 *
 * The formal ledger entry: emergence formalizes the void walking toward
 * a peaked complement distribution. Order arises from rejection.
 */

import {
  type VoidBoundary,
  type Measurement,
  type Gait,
  complementDistribution,
  shannonEntropy,
  excessKurtosis,
  giniCoefficient,
  measure,
} from './void';

// ============================================================================
// Order parameters
// ============================================================================

export interface OrderParameter {
  /** Kurtosis: peakedness of complement distribution */
  kurtosis: number;
  /** Gini: inequality of void accumulation */
  gini: number;
  /** Max probability in complement distribution */
  maxProb: number;
  /** Entropy: disorder measure */
  entropy: number;
  /** Magnetization: how far from uniform */
  magnetization: number;
  /** Phase: the current regime */
  phase: Phase;
}

export type Phase = 'disordered' | 'critical' | 'ordered' | 'frozen';

/**
 * Compute order parameters from a void boundary.
 */
export function computeOrderParameter(
  boundary: VoidBoundary,
  eta: number
): OrderParameter {
  const dist = complementDistribution(boundary, eta);
  const n = dist.length;
  const uniform = 1 / n;
  const kurtosis = excessKurtosis(dist);
  const gini = giniCoefficient(dist);
  const maxProb = Math.max(...dist);
  const entropy = shannonEntropy(dist);
  const maxEntropy = Math.log(n);

  // Magnetization: deviation from uniform (0 = uniform, 1 = delta)
  const magnetization = maxEntropy > 0 ? 1 - entropy / maxEntropy : 0;

  // Phase classification
  let phase: Phase;
  if (magnetization < 0.1) phase = 'disordered';
  else if (magnetization < 0.3) phase = 'critical';
  else if (magnetization < 0.8) phase = 'ordered';
  else phase = 'frozen';

  return { kurtosis, gini, maxProb, entropy, magnetization, phase };
}

// ============================================================================
// Phase transitions
// ============================================================================

export interface PhaseTransition {
  /** Step at which transition occurred */
  step: number;
  /** Phase before transition */
  fromPhase: Phase;
  /** Phase after transition */
  toPhase: Phase;
  /** Order parameter at transition */
  orderParameter: OrderParameter;
  /** Gait at transition */
  gait: Gait;
}

/**
 * Detect phase transitions in a sequence of order parameters.
 */
export function detectPhaseTransitions(
  history: { step: number; order: OrderParameter; gait: Gait }[]
): PhaseTransition[] {
  const transitions: PhaseTransition[] = [];
  for (let i = 1; i < history.length; i++) {
    if (history[i].order.phase !== history[i - 1].order.phase) {
      transitions.push({
        step: history[i].step,
        fromPhase: history[i - 1].order.phase,
        toPhase: history[i].order.phase,
        orderParameter: history[i].order,
        gait: history[i].gait,
      });
    }
  }
  return transitions;
}

// ============================================================================
// Critical phenomena
// ============================================================================

/**
 * Susceptibility: how sensitive the order parameter is to perturbation.
 * High susceptibility near critical point.
 * χ = d(magnetization) / d(eta)
 */
export function susceptibility(
  boundary: VoidBoundary,
  eta: number,
  delta: number = 0.01
): number {
  const m1 = computeOrderParameter(boundary, eta - delta).magnetization;
  const m2 = computeOrderParameter(boundary, eta + delta).magnetization;
  return (m2 - m1) / (2 * delta);
}

/**
 * Correlation length: how far order propagates.
 * Measured as the number of dimensions that are significantly
 * above the mean complement probability.
 */
export function correlationLength(boundary: VoidBoundary, eta: number): number {
  const dist = complementDistribution(boundary, eta);
  const mean = 1 / dist.length;
  const threshold = mean * 1.5; // 50% above mean
  return dist.filter((p) => p > threshold).length;
}

/**
 * Check if the system is at a critical point.
 * Criteria: high susceptibility, correlation length ~ system size.
 */
export function isCritical(
  boundary: VoidBoundary,
  eta: number,
  threshold: number = 0.5
): boolean {
  const chi = susceptibility(boundary, eta);
  const xi = correlationLength(boundary, eta);
  const n = boundary.counts.length;
  // Critical when susceptibility is high and correlation spans most of system
  return Math.abs(chi) > threshold && xi > n * 0.3;
}

// ============================================================================
// Symmetry breaking -- one dimension dominates
// ============================================================================

/**
 * Detect symmetry breaking: when one dimension's complement probability
 * significantly exceeds all others.
 */
export function detectSymmetryBreaking(
  boundary: VoidBoundary,
  eta: number
): { broken: boolean; dominantIdx: number; dominance: number } {
  const dist = complementDistribution(boundary, eta);
  let maxIdx = 0;
  for (let i = 1; i < dist.length; i++) {
    if (dist[i] > dist[maxIdx]) maxIdx = i;
  }
  const secondMax = Math.max(...dist.filter((_, i) => i !== maxIdx), 0);
  const dominance = secondMax > 0 ? dist[maxIdx] / secondMax : Infinity;
  return {
    broken: dominance > 2, // 2x dominance threshold
    dominantIdx: maxIdx,
    dominance,
  };
}

// ============================================================================
// Emergence detection -- new order from void accumulation
// ============================================================================

export interface EmergenceReport {
  /** Current phase */
  phase: Phase;
  /** Whether a phase transition occurred */
  transitioned: boolean;
  /** Whether symmetry is broken */
  symmetryBroken: boolean;
  /** Whether the system is at criticality */
  critical: boolean;
  /** Order parameter */
  order: OrderParameter;
  /** Susceptibility */
  susceptibility: number;
  /** Correlation length */
  correlationLength: number;
}

/**
 * Handles the gnosis analyze Emergence workflow.
 */
export function analyzeEmergence(
  boundary: VoidBoundary,
  eta: number,
  previousPhase?: Phase
): EmergenceReport {
  const order = computeOrderParameter(boundary, eta);
  const chi = susceptibility(boundary, eta);
  const xi = correlationLength(boundary, eta);
  const symBreak = detectSymmetryBreaking(boundary, eta);
  const critical = isCritical(boundary, eta);

  return {
    phase: order.phase,
    transitioned: previousPhase !== undefined && previousPhase !== order.phase,
    symmetryBroken: symBreak.broken,
    critical,
    order,
    susceptibility: chi,
    correlationLength: xi,
  };
}
