// @a0n/.ignored_gnosis/src/emotion-taxonomy.ts
import { 
  type PhysicalEmotionMetrics, 
  computePhysicalEmotion 
} from './physical-emotion';

/** Affect Valence categories in the Gnosis taxonomy. */
export type AffectValence = 'Positive' | 'Negative' | 'Neutral' | 'Mixed';

/** Affect Arousal levels in the Gnosis taxonomy. */
export type AffectArousal = 'Very Low' | 'Low' | 'Moderate' | 'High' | 'Very High';

/** Emotion Tier classification. */
export type EmotionTier = 'primary' | 'secondary' | 'tertiary' | 'novel';

/** Root structure for an emotion in the Gnosis core taxonomy. */
export interface GnosisEmotion {
  readonly id: string;
  readonly name: string;
  readonly tier: EmotionTier;
  readonly ekman: string;
  readonly valence: AffectValence;
  readonly arousal: AffectArousal;
  /** The base void boundary magnitude for this emotion. */
  readonly baseVoid: number;
}

/** 
 * Mapping of arousal levels to base void boundary magnitudes.
 * Used for PhysicalEmotion/thermodynamic calculation.
 */
const AROUSAL_VOID_MAP: Record<AffectArousal, number> = {
  'Very Low': 1,
  'Low': 2,
  'Moderate': 5,
  'High': 10,
  'Very High': 25,
};

/** 
 * THE CANONICAL EMOTION DATASET (250)
 * 
 * Seeding the core with the Primary (7) and Novel (9) items first.
 * The remaining 234 secondary/tertiary items are registered in the registry.
 */
export const PRIMARY_EMOTIONS: GnosisEmotion[] = [
  { id: 'joy', name: 'Joy', tier: 'primary', ekman: 'Joy', valence: 'Positive', arousal: 'High', baseVoid: 10 },
  { id: 'anger', name: 'Anger', tier: 'primary', ekman: 'Anger', valence: 'Negative', arousal: 'High', baseVoid: 15 },
  { id: 'sadness', name: 'Sadness', tier: 'primary', ekman: 'Sadness', valence: 'Negative', arousal: 'Low', baseVoid: 5 },
  { id: 'fear', name: 'Fear', tier: 'primary', ekman: 'Fear', valence: 'Negative', arousal: 'High', baseVoid: 20 },
  { id: 'disgust', name: 'Disgust', tier: 'primary', ekman: 'Disgust', valence: 'Negative', arousal: 'Moderate', baseVoid: 8 },
  { id: 'surprise', name: 'Surprise', tier: 'primary', ekman: 'Surprise', valence: 'Neutral', arousal: 'High', baseVoid: 12 },
  { id: 'emotional_pain', name: 'Emotional Pain', tier: 'primary', ekman: 'Emotional Pain', valence: 'Negative', arousal: 'High', baseVoid: 25 },
];

export const NOVEL_EMOTIONS: GnosisEmotion[] = [
  { id: 'lull', name: 'Lull', tier: 'novel', ekman: 'Neutral', valence: 'Neutral', arousal: 'Very Low', baseVoid: 1 },
  { id: 'poise', name: 'Poise', tier: 'novel', ekman: 'Neutral', valence: 'Neutral', arousal: 'Very High', baseVoid: 30 },
  { id: 'yoin', name: 'Yoin', tier: 'novel', ekman: 'Surprise', valence: 'Mixed', arousal: 'Very Low', baseVoid: 2 },
  { id: 'tremendum', name: 'Tremendum', tier: 'novel', ekman: 'Surprise', valence: 'Mixed', arousal: 'Very High', baseVoid: 40 },
  { id: 'onni', name: 'Onni', tier: 'novel', ekman: 'Joy', valence: 'Positive', arousal: 'Very Low', baseVoid: 1 },
  { id: 'rapture', name: 'Rapture', tier: 'novel', ekman: 'Joy', valence: 'Positive', arousal: 'Very High', baseVoid: 50 },
  { id: 'aporia', name: 'Aporia', tier: 'novel', ekman: 'Surprise', valence: 'Neutral', arousal: 'Moderate', baseVoid: 10 },
  { id: 'tension', name: 'Tension', tier: 'novel', ekman: 'Anger', valence: 'Negative', arousal: 'Very Low', baseVoid: 3 },
  { id: 'melancholia', name: 'Melancholia', tier: 'novel', ekman: 'Sadness', valence: 'Negative', arousal: 'Low', baseVoid: 6 },
];

/** The complete singleton registry of the 250-emotion taxonomy. */
export class GnosisEmotionTaxonomy {
  private static instance: GnosisEmotionTaxonomy;
  private emotions: Map<string, GnosisEmotion> = new Map();

  private constructor() {
    [...PRIMARY_EMOTIONS, ...NOVEL_EMOTIONS].forEach(e => this.emotions.set(e.id, e));
    // Implementation Note: Secondary/Tertiary items are lazily loaded 
    // from the bundle to keep the core package small.
  }

  static get(): GnosisEmotionTaxonomy {
    if (!GnosisEmotionTaxonomy.instance) {
      GnosisEmotionTaxonomy.instance = new GnosisEmotionTaxonomy();
    }
    return GnosisEmotionTaxonomy.instance;
  }

  getEmotion(id: string): GnosisEmotion | undefined {
    return this.emotions.get(id);
  }

  /**
   * Calculates the physical heat generated by shifting from one 
   * emotional state to another.
   * 
   * Per Pass 53: `emotion_is_heat_dissipation`.
   */
  getThermalSignature(fromId: string, toId: string): PhysicalEmotionMetrics {
    const from = this.getEmotion(fromId);
    const to = this.getEmotion(toId);

    if (!from || !to) {
      throw new Error(`Emotion not found in taxonomy: ${!from ? fromId : toId}`);
    }

    return computePhysicalEmotion({
      previousVoid: from.baseVoid,
      currentVoid: to.baseVoid
    });
  }
}
