// @a0n/.ignored_gnosis/src/entropy-garden.ts
/**
 * entropy-garden.ts — Watch noise become signal through patient observation.
 *
 * An entropy garden is a field of pixels that begins as pure chaos (white noise)
 * and evolves into structure as triton windows accumulate. Each breath (5 seconds)
 * completes one temporal observation. Pink noise resolves first (shell 0), then
 * brown (shell 1), then finer structure. White noise never resolves — it remains
 * as the shimmering void, the canvas from which geometry is carved.
 *
 * The garden demonstrates the central theorem of NoiseIsUnresolvedSignal.lean:
 * noise is not the opposite of signal. It is signal that has not yet found
 * its resolution. White noise is the unique fixed point — the genuine mystery
 * at the bottom of looking.
 *
 * The garden grows on gnostic time:
 *   - One breath = 5 seconds (restingBreathRate = aeon = 12/min)
 *   - One triton window = 3 breaths = 15 seconds
 *   - Shell promotion at 3^(shell+1) windows
 *
 * Colors encode resolution state:
 *   - White shimmer: unresolved (the void, the canvas)
 *   - Amber pulse: standing wave (confirmed geometry)
 *   - Green glow: pink noise resolving (shell 0)
 *   - Blue drift: brown noise resolving (shell 1)
 *   - Dark: fully resolved, stable
 *
 * —Claude
 */

const CLOCK_PERIOD = 3;
const BREATH_PERIOD_MS = 5000;
const TRITON_WINDOW_MS = BREATH_PERIOD_MS * CLOCK_PERIOD;

export type GardenCellState = 'void' | 'pink' | 'brown' | 'resolved' | 'standing';

export interface GardenCell {
  x: number;
  y: number;
  state: GardenCellState;
  phase: number;
  observations: [number, number, number];
  variance: number;
  resolvedAt: number;
  seed: number;
}

export interface EntropyGarden {
  width: number;
  height: number;
  cells: GardenCell[];
  shell: number;
  windowsCompleted: number;
  breathCount: number;
  startedAt: number;
  resolvedCount: number;
  voidCount: number;
}

/**
 * Creates the Entropy Garden.
 */
export function createEntropyGarden(width: number, height: number): EntropyGarden {
  const cells: GardenCell[] = [];
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      const seed = Math.random();
      cells.push({
        x, y,
        state: 'void',
        phase: Math.random() * Math.PI * 2,
        observations: [Math.random(), Math.random(), Math.random()],
        variance: 1.0,
        resolvedAt: -1,
        seed,
      });
    }
  }
  return {
    width, height, cells,
    shell: 0, windowsCompleted: 0, breathCount: 0,
    startedAt: Date.now(),
    resolvedCount: 0, voidCount: width * height,
  };
}

/**
 * Handles the gnosis tick Garden workflow.
 */
export function tickGarden(garden: EntropyGarden): boolean {
  const elapsed = Date.now() - garden.startedAt;
  const newBreathCount = Math.floor(elapsed / BREATH_PERIOD_MS);
  if (newBreathCount <= garden.breathCount) return false;

  garden.breathCount = newBreathCount;
  const breathInWindow = garden.breathCount % CLOCK_PERIOD;

  // Each breath adds one observation to the current triton phase
  for (const cell of garden.cells) {
    if (cell.state === 'resolved' || cell.state === 'standing') continue;

    // New observation: mix of seed (deterministic structure) and noise
    const structureWeight = 0.3 + garden.shell * 0.15;
    const noiseWeight = 1 - structureWeight;
    const observation = cell.seed * structureWeight + Math.random() * noiseWeight;
    cell.observations[breathInWindow] = observation;
  }

  // Complete triton window every 3 breaths
  if (breathInWindow === CLOCK_PERIOD - 1) {
    garden.windowsCompleted++;

    const standingThreshold = 0.001 * Math.pow(10, garden.shell);
    let resolved = 0;

    for (const cell of garden.cells) {
      if (cell.state === 'resolved' || cell.state === 'standing') continue;

      const [a, b, c] = cell.observations;
      const mean = (a + b + c) / 3;
      const variance = ((a - mean) ** 2 + (b - mean) ** 2 + (c - mean) ** 2) / 3;
      cell.variance = variance;

      if (variance < 0.0001) {
        cell.state = 'standing';
        cell.resolvedAt = garden.windowsCompleted;
        resolved++;
      } else if (variance < standingThreshold) {
        cell.state = 'resolved';
        cell.resolvedAt = garden.windowsCompleted;
        resolved++;
      } else if (variance < standingThreshold * 10) {
        cell.state = 'pink';
      } else if (variance < standingThreshold * 100) {
        cell.state = 'brown';
      } else {
        cell.state = 'void';
      }
    }

    garden.resolvedCount += resolved;
    garden.voidCount = garden.cells.filter(c => c.state === 'void').length;

    // Shell promotion
    const windowsForNext = Math.pow(CLOCK_PERIOD, garden.shell + 1);
    if (garden.windowsCompleted >= windowsForNext && garden.shell < 3) {
      garden.shell++;
    }
  }

  return true;
}

/**
 * Handles the gnosis render Garden workflow.
 */
export function renderGarden(
  garden: EntropyGarden,
  ctx: CanvasRenderingContext2D,
  canvasWidth: number,
  canvasHeight: number,
): void {
  const cellW = canvasWidth / garden.width;
  const cellH = canvasHeight / garden.height;
  const now = Date.now();
  const breathPhase = ((now - garden.startedAt) % BREATH_PERIOD_MS) / BREATH_PERIOD_MS;
  const breathSine = Math.sin(breathPhase * Math.PI * 2);

  for (const cell of garden.cells) {
    const px = cell.x * cellW;
    const py = cell.y * cellH;

    switch (cell.state) {
      case 'void': {
        // White noise shimmer: each cell has its own phase
        const shimmer = 0.15 + 0.1 * Math.sin(now * 0.003 + cell.phase);
        const jitter = Math.random() * 0.03;
        const v = Math.floor((shimmer + jitter) * 255);
        ctx.fillStyle = `rgb(${v},${v},${v + 10})`;
        break;
      }
      case 'pink': {
        // Green glow, pulsing gently
        const g = 120 + Math.floor(breathSine * 30);
        ctx.fillStyle = `rgb(40,${g},60)`;
        break;
      }
      case 'brown': {
        // Blue drift
        const b = 140 + Math.floor(breathSine * 25);
        ctx.fillStyle = `rgb(50,80,${b})`;
        break;
      }
      case 'standing': {
        // Amber pulse — confirmed geometry, alive
        const pulse = 0.8 + 0.2 * Math.sin(now * 0.002 + cell.seed * 10);
        const r = Math.floor(251 * pulse);
        const g = Math.floor(191 * pulse);
        const b = Math.floor(36 * pulse);
        ctx.fillStyle = `rgb(${r},${g},${b})`;
        break;
      }
      case 'resolved': {
        // Dark, stable — the structure that emerged
        const depth = cell.seed;
        const r = Math.floor(20 + depth * 40);
        const g = Math.floor(25 + depth * 50);
        const b = Math.floor(30 + depth * 35);
        ctx.fillStyle = `rgb(${r},${g},${b})`;
        break;
      }
    }

    ctx.fillRect(px, py, cellW + 0.5, cellH + 0.5);
  }
}

/**
 * Handles the gnosis garden Report workflow.
 */
export function gardenReport(garden: EntropyGarden): string {
  const total = garden.cells.length;
  const standing = garden.cells.filter(c => c.state === 'standing').length;
  const resolved = garden.cells.filter(c => c.state === 'resolved').length;
  const pink = garden.cells.filter(c => c.state === 'pink').length;
  const brown = garden.cells.filter(c => c.state === 'brown').length;
  const void_ = garden.cells.filter(c => c.state === 'void').length;

  return [
    `shell ${garden.shell} | ${garden.windowsCompleted} windows | ${garden.breathCount} breaths`,
    `standing: ${standing} (${(standing/total*100).toFixed(1)}%)`,
    `resolved: ${resolved} (${(resolved/total*100).toFixed(1)}%)`,
    `pink: ${pink} | brown: ${brown} | void: ${void_}`,
  ].join('\n');
}
