// @a0n/.ignored_gnosis/src/eyetracking.ts
/**
 * eyetracking.ts -- Gnosis-native eye-tracking corpus and failure semantics
 *
 * Eye tracking is treated as an append-only rejection surface:
 *   - observations are the witness stream
 *   - calibration is a finite cave with explicit deficit
 *   - failures are vents that grow a VoidBoundary
 *   - the failure log is Merkle-verifiable
 *   - the session is replayable through QDoc
 *
 * This makes eye-tracking failures first-class training data for Dash,
 * Neural, void-os, and downstream ETaaS services without introducing a
 * second semantics layer outside gnosis.
 */

import { QDoc } from './crdt/qdoc';
import {
  fullBuleyeanAudit,
  type FullAuditReport,
} from './seven-laws-checker';
import {
  buildVoidMerkleTree,
  type VoidMerkleTree,
  type VoidVentRecord,
} from './void-merkle';
import {
  createVoidBoundary,
  updateVoidBoundary,
  type VoidBoundary,
} from './void';

export interface GazePoint {
  x: number;
  y: number;
}

export interface EyeTrackingEyeReading {
  pupil: GazePoint;
  openness: number;
  quality: number;
}

export const EYE_TRACKING_ZONES = [
  'top-left',
  'top-center',
  'top-right',
  'center-left',
  'center',
  'center-right',
  'bottom-left',
  'bottom-center',
  'bottom-right',
  'unavailable',
] as const;

export type EyeTrackingZone = (typeof EYE_TRACKING_ZONES)[number];

export interface EyeTrackingObservation {
  observationId: string;
  timestamp: number;
  left: EyeTrackingEyeReading | null;
  right: EyeTrackingEyeReading | null;
  gazePoint: GazePoint | null;
  zone: EyeTrackingZone;
  confidence: number | null;
  blink: boolean;
  eyeBlocked: boolean;
  fixationDurationMs: number;
  saccadeVelocity: number;
  scanningVariance: number;
  faceQuality?: number;
  source?: string;
  metadata?: Record<string, unknown>;
}

export interface EyeTrackingCalibrationPoint extends GazePoint {
  id: string;
}

export interface EyeTrackingCalibrationPlan {
  points: EyeTrackingCalibrationPoint[];
  samplesPerPoint: number;
  maxAttemptsPerPoint: number;
}

export const EYE_TRACKING_FAILURE_DIMENSIONS = [
  'calibration-deficit',
  'calibration-drift',
  'missing-eyes',
  'low-quality',
  'pupil-instability',
  'false-blink',
  'false-eye-block',
  'zone-miss',
  'confidence-overstatement',
  'reacquisition-failure',
  'lighting-instability',
  'occlusion',
] as const;

export type EyeTrackingFailureKind =
  (typeof EYE_TRACKING_FAILURE_DIMENSIONS)[number];

export type EyeTrackingSeverity = 'info' | 'warning' | 'error' | 'fatal';

export type EyeTrackingCalibrationRejectionReason =
  | EyeTrackingFailureKind
  | 'point-complete'
  | 'point-exhausted'
  | 'unknown-point';

export interface EyeTrackingCalibrationSample {
  sampleId: string;
  sessionId: string;
  pointId: string;
  point: GazePoint;
  observationId: string;
  timestamp: number;
  quality: number;
  outcome: 'accepted' | 'rejected';
  rejectionReason?: EyeTrackingCalibrationRejectionReason;
}

export interface EyeTrackingCalibrationPointStatus {
  pointId: string;
  accepted: number;
  rejected: number;
  observed: number;
  deficit: number;
  remainingAttempts: number;
  complete: boolean;
  exhausted: boolean;
}

export interface EyeTrackingCalibrationLedger {
  plan: EyeTrackingCalibrationPlan;
  requiredSamples: number;
  acceptedSamples: number;
  rejectedSamples: number;
  observedSamples: number;
  deficit: number;
  ready: boolean;
  terminated: boolean;
  nextPointId: string | null;
  points: EyeTrackingCalibrationPointStatus[];
}

export interface EyeTrackingFailureEvent {
  failureId: string;
  sessionId: string;
  timestamp: number;
  kind: EyeTrackingFailureKind;
  magnitude: number;
  severity: EyeTrackingSeverity;
  message: string;
  observationId?: string;
  expectedZone?: EyeTrackingZone;
  observedZone?: EyeTrackingZone;
  metadata?: Record<string, unknown>;
}

export type EyeTrackingSessionStatus =
  | 'calibrating'
  | 'tracking'
  | 'completed'
  | 'failed';

export interface EyeTrackingSession {
  sessionId: string;
  startedAt: number;
  endedAt?: number;
  status: EyeTrackingSessionStatus;
  deviceId?: string;
  modelVersion?: string;
  calibration: EyeTrackingCalibrationLedger;
  observations: EyeTrackingObservation[];
  calibrationSamples: EyeTrackingCalibrationSample[];
  failures: EyeTrackingFailureEvent[];
  boundary: VoidBoundary;
  metadata: Record<string, unknown>;
}

export interface EyeTrackingCorpus {
  sessions: EyeTrackingSession[];
  boundary: VoidBoundary;
}

export interface EyeTrackingInvariant {
  id: string;
  theoremId: string;
  pass: boolean;
  details: string;
}

export interface EyeTrackingAuditReport {
  pass: boolean;
  buleyean: FullAuditReport;
  invariants: EyeTrackingInvariant[];
  merkle: {
    rootHash: number;
    size: number;
  };
}

export interface EyeTrackingRejectionRecord {
  prompt: string;
  rejectedResponses: string[];
  rejectionCounts: number[];
  totalRounds: number;
  metadata?: Record<string, unknown>;
}

export interface EyeTrackingFailurePressure {
  sessionCount: number;
  totalMagnitude: number;
  byKind: Record<EyeTrackingFailureKind, number>;
  byDevice: Record<string, number>;
  byModel: Record<string, number>;
  hottestKind: EyeTrackingFailureKind | null;
}

export type EyeTrackingDatasetSplit =
  | 'train'
  | 'eval'
  | 'shadow'
  | 'holdout';

export interface EyeTrackingDatasetSelection {
  deviceIds?: readonly string[];
  modelVersions?: readonly string[];
  statuses?: readonly EyeTrackingSessionStatus[];
  requireReadyCalibration?: boolean;
  failureKinds?: readonly EyeTrackingFailureKind[];
  minTotalFailureMagnitude?: number;
  maxSessions?: number;
  sortBy?: 'startedAt' | 'failureMagnitude';
}

export interface EyeTrackingMerkleWitness {
  sessionId: string;
  rootHash: number;
  size: number;
}

export interface EyeTrackingDatasetSlice {
  sliceId: string;
  split: EyeTrackingDatasetSplit;
  createdAt: number;
  sessionIds: string[];
  sessionCount: number;
  sessions: EyeTrackingSession[];
  totalObservations: number;
  totalFailures: number;
  totalCalibrationDeficit: number;
  boundary: VoidBoundary;
  pressure: EyeTrackingFailurePressure;
  merkle: EyeTrackingMerkleWitness[];
  rejectionJsonl: string;
  metadata: Record<string, unknown>;
}

export interface EyeTrackingTrainingSplitInput {
  split: EyeTrackingDatasetSplit;
  source:
    | EyeTrackingCorpus
    | EyeTrackingSession
    | readonly EyeTrackingSession[];
  selection?: EyeTrackingDatasetSelection;
  metadata?: Record<string, unknown>;
}

export interface EyeTrackingTrainingManifest {
  manifestId: string;
  service: 'eyetracking-as-a-service';
  generatedAt: number;
  slices: EyeTrackingDatasetSlice[];
  totalSessionCount: number;
  totalObservationCount: number;
  totalFailureCount: number;
  totalCalibrationDeficit: number;
  boundary: VoidBoundary;
  hottestKinds: EyeTrackingFailureKind[];
  metadata: Record<string, unknown>;
}

export type EyeTrackingBuleyeanCurriculum =
  | 'inverse_bule'
  | 'kurtosis'
  | 'rejection_density'
  | 'random';

export interface EyeTrackingBuleyeanFederatedConfig {
  dimensions: number;
  lr: number;
  minRejections: number;
  maxGradNorm: number;
  sparsityThreshold: number;
  nodeId: string;
}

export interface EyeTrackingBuleyeanTrainingConfig {
  alpha: number;
  contrastWeight: number;
  sftWeight: number;
  temperature: number;
  minRejectionCount: number;
  sparse: boolean;
  curriculum: EyeTrackingBuleyeanCurriculum;
  federated: EyeTrackingBuleyeanFederatedConfig;
}

export interface EyeTrackingBuleyeanTrainingConfigOverrides
  extends Partial<
    Omit<EyeTrackingBuleyeanTrainingConfig, 'federated'>
  > {
  federated?: Partial<EyeTrackingBuleyeanFederatedConfig>;
}

export interface EyeTrackingBuleyeanTrainingJobInput {
  split: EyeTrackingDatasetSplit;
  sessions: EyeTrackingSession[];
  selection?: EyeTrackingDatasetSelection;
  metadata?: Record<string, unknown>;
}

export interface EyeTrackingBuleyeanSplitSummary {
  split: EyeTrackingDatasetSplit;
  sessionIds: string[];
  sessionCount: number;
  recordCount: number;
  totalRounds: number;
  totalRejectionMass: number;
}

export interface EyeTrackingBuleyeanCorpusSummary {
  promptCount: number;
  sessionCount: number;
  totalRejectedResponses: number;
  totalRejectionMass: number;
  totalRounds: number;
  totalObservations: number;
  totalCalibrationDeficit: number;
  averageRejectedResponsesPerPrompt: number;
  averageRejectionMassPerPrompt: number;
  byKind: Record<EyeTrackingFailureKind, number>;
  byDevice: Record<string, number>;
  byModel: Record<string, number>;
  byStatus: Record<EyeTrackingSessionStatus, number>;
  hottestKind: EyeTrackingFailureKind | null;
}

export interface EyeTrackingBuleyeanTrainingJob {
  jobId: string;
  createdAt: number;
  service: 'eyetracking-as-a-service';
  trainer: 'buleyean-rl';
  inputs: EyeTrackingBuleyeanTrainingJobInput[];
  manifest: EyeTrackingTrainingManifest;
  splits: EyeTrackingBuleyeanSplitSummary[];
  records: EyeTrackingRejectionRecord[];
  rejectionJsonl: string;
  corpus: EyeTrackingBuleyeanCorpusSummary;
  config: EyeTrackingBuleyeanTrainingConfig;
  metadata: Record<string, unknown>;
}

export interface EyeTrackingBuleyeanAppendResult {
  job: EyeTrackingBuleyeanTrainingJob;
  appendedSessionIds: string[];
  duplicateSessionIds: string[];
}

export interface EyeTrackingBuleyeanPersonalizationDelta {
  deltaId: string;
  createdAt: number;
  trainer: 'buleyean-rl';
  sessionIds: string[];
  sessionCount: number;
  records: EyeTrackingRejectionRecord[];
  rejectionJsonl: string;
  corpus: EyeTrackingBuleyeanCorpusSummary;
  selection?: EyeTrackingDatasetSelection;
  metadata: Record<string, unknown>;
}

export interface EyeTrackingBuleyeanTrainingJobOptions {
  createdAt?: number;
  config?: EyeTrackingBuleyeanTrainingConfigOverrides;
  metadata?: Record<string, unknown>;
}

export interface EyeTrackingBuleyeanAppendOptions {
  createdAt?: number;
  metadata?: Record<string, unknown>;
}

export interface EyeTrackingBuleyeanCorpusSummaryOptions {
  selection?: EyeTrackingDatasetSelection;
}

export interface EyeTrackingBuleyeanPersonalizationDeltaOptions {
  createdAt?: number;
  selection?: EyeTrackingDatasetSelection;
  metadata?: Record<string, unknown>;
}

const SESSION_MAP = 'eyetracking.sessions';
const SESSION_ALIAS_MAP = 'eyetracking.aliases';
const CURRENT_SESSION_ALIAS = 'current';

const FAILURE_KIND_TO_DIMENSION = new Map<EyeTrackingFailureKind, number>(
  EYE_TRACKING_FAILURE_DIMENSIONS.map((kind, index) => [kind, index])
);

const SEVERITY_TO_MAGNITUDE: Record<EyeTrackingSeverity, number> = {
  info: 1,
  warning: 2,
  error: 3,
  fatal: 5,
};

function clonePoint(point: GazePoint): GazePoint {
  return { x: point.x, y: point.y };
}

function stableStringify(value: unknown): string {
  if (value === null || typeof value !== 'object') {
    return JSON.stringify(value);
  }

  if (Array.isArray(value)) {
    return `[${value.map((entry) => stableStringify(entry)).join(',')}]`;
  }

  const entries = Object.entries(value as Record<string, unknown>)
    .filter(([, entryValue]) => entryValue !== undefined)
    .sort(([left], [right]) => left.localeCompare(right))
    .map(
      ([key, entryValue]) =>
        `${JSON.stringify(key)}:${stableStringify(entryValue)}`
    );

  return `{${entries.join(',')}}`;
}

function createStableHash(source: string): string {
  let hash = 2166136261;
  for (let index = 0; index < source.length; index++) {
    hash ^= source.charCodeAt(index);
    hash = Math.imul(hash, 16777619);
  }

  return (hash >>> 0).toString(16).padStart(8, '0');
}

function cloneEyeReading(
  reading: EyeTrackingEyeReading | null
): EyeTrackingEyeReading | null {
  if (reading === null) {
    return null;
  }

  return {
    pupil: clonePoint(reading.pupil),
    openness: reading.openness,
    quality: reading.quality,
  };
}

function cloneVoidBoundary(boundary: VoidBoundary): VoidBoundary {
  return {
    counts: [...boundary.counts],
    totalEntries: boundary.totalEntries,
  };
}

function cloneSelection(
  selection: EyeTrackingDatasetSelection
): EyeTrackingDatasetSelection;
function cloneSelection(
  selection: EyeTrackingDatasetSelection | undefined
): EyeTrackingDatasetSelection | undefined {
  if (selection === undefined) {
    return undefined;
  }

  return {
    ...(selection.deviceIds !== undefined
      ? { deviceIds: [...selection.deviceIds] }
      : {}),
    ...(selection.modelVersions !== undefined
      ? { modelVersions: [...selection.modelVersions] }
      : {}),
    ...(selection.statuses !== undefined
      ? { statuses: [...selection.statuses] }
      : {}),
    ...(selection.requireReadyCalibration !== undefined
      ? { requireReadyCalibration: selection.requireReadyCalibration }
      : {}),
    ...(selection.failureKinds !== undefined
      ? { failureKinds: [...selection.failureKinds] }
      : {}),
    ...(selection.minTotalFailureMagnitude !== undefined
      ? {
          minTotalFailureMagnitude: selection.minTotalFailureMagnitude,
        }
      : {}),
    ...(selection.maxSessions !== undefined
      ? { maxSessions: selection.maxSessions }
      : {}),
    ...(selection.sortBy !== undefined ? { sortBy: selection.sortBy } : {}),
  };
}

function cloneCalibrationPlan(
  plan: EyeTrackingCalibrationPlan
): EyeTrackingCalibrationPlan {
  return {
    points: plan.points.map((point) => ({
      id: point.id,
      x: point.x,
      y: point.y,
    })),
    samplesPerPoint: plan.samplesPerPoint,
    maxAttemptsPerPoint: plan.maxAttemptsPerPoint,
  };
}

function cloneObservation(
  observation: EyeTrackingObservation
): EyeTrackingObservation {
  return {
    observationId: observation.observationId,
    timestamp: observation.timestamp,
    left: cloneEyeReading(observation.left),
    right: cloneEyeReading(observation.right),
    gazePoint:
      observation.gazePoint === null ? null : clonePoint(observation.gazePoint),
    zone: observation.zone,
    confidence: observation.confidence,
    blink: observation.blink,
    eyeBlocked: observation.eyeBlocked,
    fixationDurationMs: observation.fixationDurationMs,
    saccadeVelocity: observation.saccadeVelocity,
    scanningVariance: observation.scanningVariance,
    ...(observation.faceQuality !== undefined
      ? { faceQuality: observation.faceQuality }
      : {}),
    ...(observation.source !== undefined ? { source: observation.source } : {}),
    ...(observation.metadata !== undefined
      ? { metadata: { ...observation.metadata } }
      : {}),
  };
}

function cloneCalibrationSample(
  sample: EyeTrackingCalibrationSample
): EyeTrackingCalibrationSample {
  return {
    sampleId: sample.sampleId,
    sessionId: sample.sessionId,
    pointId: sample.pointId,
    point: clonePoint(sample.point),
    observationId: sample.observationId,
    timestamp: sample.timestamp,
    quality: sample.quality,
    outcome: sample.outcome,
    ...(sample.rejectionReason !== undefined
      ? { rejectionReason: sample.rejectionReason }
      : {}),
  };
}

function cloneFailure(
  failure: EyeTrackingFailureEvent
): EyeTrackingFailureEvent {
  return {
    failureId: failure.failureId,
    sessionId: failure.sessionId,
    timestamp: failure.timestamp,
    kind: failure.kind,
    magnitude: failure.magnitude,
    severity: failure.severity,
    message: failure.message,
    ...(failure.observationId !== undefined
      ? { observationId: failure.observationId }
      : {}),
    ...(failure.expectedZone !== undefined
      ? { expectedZone: failure.expectedZone }
      : {}),
    ...(failure.observedZone !== undefined
      ? { observedZone: failure.observedZone }
      : {}),
    ...(failure.metadata !== undefined
      ? { metadata: { ...failure.metadata } }
      : {}),
  };
}

function cloneSession(session: EyeTrackingSession): EyeTrackingSession {
  return {
    sessionId: session.sessionId,
    startedAt: session.startedAt,
    ...(session.endedAt !== undefined ? { endedAt: session.endedAt } : {}),
    status: session.status,
    ...(session.deviceId !== undefined ? { deviceId: session.deviceId } : {}),
    ...(session.modelVersion !== undefined
      ? { modelVersion: session.modelVersion }
      : {}),
    calibration: cloneCalibrationLedger(session.calibration),
    observations: session.observations.map(cloneObservation),
    calibrationSamples: session.calibrationSamples.map(cloneCalibrationSample),
    failures: session.failures.map(cloneFailure),
    boundary: cloneVoidBoundary(session.boundary),
    metadata: { ...session.metadata },
  };
}

function cloneCalibrationLedger(
  ledger: EyeTrackingCalibrationLedger
): EyeTrackingCalibrationLedger {
  return {
    plan: cloneCalibrationPlan(ledger.plan),
    requiredSamples: ledger.requiredSamples,
    acceptedSamples: ledger.acceptedSamples,
    rejectedSamples: ledger.rejectedSamples,
    observedSamples: ledger.observedSamples,
    deficit: ledger.deficit,
    ready: ledger.ready,
    terminated: ledger.terminated,
    nextPointId: ledger.nextPointId,
    points: ledger.points.map((point) => ({ ...point })),
  };
}

function createEmptyFailureCounts(): Record<EyeTrackingFailureKind, number> {
  return {
    'calibration-deficit': 0,
    'calibration-drift': 0,
    'missing-eyes': 0,
    'low-quality': 0,
    'pupil-instability': 0,
    'false-blink': 0,
    'false-eye-block': 0,
    'zone-miss': 0,
    'confidence-overstatement': 0,
    'reacquisition-failure': 0,
    'lighting-instability': 0,
    occlusion: 0,
  };
}

function isFailureKind(
  value: EyeTrackingCalibrationRejectionReason | undefined
): value is EyeTrackingFailureKind {
  return (
    value !== undefined &&
    FAILURE_KIND_TO_DIMENSION.has(value as EyeTrackingFailureKind)
  );
}

function normalizeMagnitude(
  magnitude: number | undefined,
  severity: EyeTrackingSeverity
): number {
  if (typeof magnitude !== 'number' || !Number.isFinite(magnitude)) {
    return SEVERITY_TO_MAGNITUDE[severity];
  }

  return Math.max(1, Math.round(magnitude));
}

function buildCalibrationLedger(
  plan: EyeTrackingCalibrationPlan,
  samples: readonly EyeTrackingCalibrationSample[]
): EyeTrackingCalibrationLedger {
  const counts = new Map<
    string,
    { accepted: number; rejected: number; observed: number }
  >(
    plan.points.map((point) => [
      point.id,
      { accepted: 0, rejected: 0, observed: 0 },
    ])
  );

  for (const sample of samples) {
    const entry = counts.get(sample.pointId);
    if (!entry) {
      continue;
    }

    entry.observed += 1;
    if (sample.outcome === 'accepted') {
      entry.accepted += 1;
    } else {
      entry.rejected += 1;
    }
  }

  const points = plan.points.map((point) => {
    const pointCounts = counts.get(point.id) ?? {
      accepted: 0,
      rejected: 0,
      observed: 0,
    };
    const deficit = Math.max(0, plan.samplesPerPoint - pointCounts.accepted);
    const remainingAttempts = Math.max(
      0,
      plan.maxAttemptsPerPoint - pointCounts.observed
    );
    const complete = deficit === 0;
    const exhausted = !complete && remainingAttempts === 0;

    return {
      pointId: point.id,
      accepted: pointCounts.accepted,
      rejected: pointCounts.rejected,
      observed: pointCounts.observed,
      deficit,
      remainingAttempts,
      complete,
      exhausted,
    };
  });

  const acceptedSamples = points.reduce((sum, point) => sum + point.accepted, 0);
  const rejectedSamples = points.reduce((sum, point) => sum + point.rejected, 0);
  const observedSamples = points.reduce((sum, point) => sum + point.observed, 0);
  const requiredSamples = plan.points.length * plan.samplesPerPoint;
  const deficit = Math.max(0, requiredSamples - acceptedSamples);
  const terminated = points.every((point) => point.complete || point.exhausted);
  const ready = terminated && deficit === 0;
  const nextPointId =
    points.find((point) => !point.complete && !point.exhausted)?.pointId ?? null;

  return {
    plan: cloneCalibrationPlan(plan),
    requiredSamples,
    acceptedSamples,
    rejectedSamples,
    observedSamples,
    deficit,
    ready,
    terminated,
    nextPointId,
    points,
  };
}

function recomputeCalibration(session: EyeTrackingSession): void {
  session.calibration = buildCalibrationLedger(
    session.calibration.plan,
    session.calibrationSamples
  );

  if (session.status === 'calibrating' && session.calibration.ready) {
    session.status = 'tracking';
  }
}

function normalizeAlias(value: string): string {
  return value.trim();
}

function toEyeTrackingSessions(
  source:
    | EyeTrackingCorpus
    | EyeTrackingSession
    | readonly EyeTrackingSession[]
): readonly EyeTrackingSession[] {
  if (Array.isArray(source)) {
    return source;
  }
  if ('sessions' in source) {
    return source.sessions;
  }
  return [source as EyeTrackingSession];
}

function createEmptyStatusCounts(): Record<EyeTrackingSessionStatus, number> {
  return {
    calibrating: 0,
    tracking: 0,
    completed: 0,
    failed: 0,
  };
}

function cloneEyeTrackingBuleyeanTrainingJobInput(
  input: EyeTrackingBuleyeanTrainingJobInput
): EyeTrackingBuleyeanTrainingJobInput {
  const cloned: EyeTrackingBuleyeanTrainingJobInput = {
    split: input.split,
    sessions: input.sessions.map((session) => cloneSession(session)),
  };
  if (input.selection !== undefined) {
    cloned.selection = cloneSelection(input.selection);
  }
  if (input.metadata !== undefined) {
    cloned.metadata = { ...input.metadata };
  }
  return cloned;
}

function createEyeTrackingBoundarySummary(
  sessions: readonly EyeTrackingSession[]
): VoidBoundary {
  const boundary = createVoidBoundary(EYE_TRACKING_FAILURE_DIMENSIONS.length);

  for (const session of sessions) {
    for (let index = 0; index < session.boundary.counts.length; index++) {
      const magnitude = session.boundary.counts[index];
      if (typeof magnitude === 'number' && magnitude > 0) {
        updateVoidBoundary(boundary, index, magnitude);
      }
    }
  }

  return boundary;
}

function createStableSessionSignature(
  sessions: readonly EyeTrackingSession[]
): string {
  const source = sessions
    .map((session) => {
      const merkle = buildEyeTrackingMerkleTree(session);
      return `${session.sessionId}:${session.startedAt}:${merkle.rootHash}:${merkle.size}`;
    })
    .sort()
    .join('|');

  return createStableHash(source);
}

/**
 * Handles the gnosis default Eye Tracking Calibration Plan workflow.
 */
export function defaultEyeTrackingCalibrationPlan(): EyeTrackingCalibrationPlan {
  return {
    points: [
      { id: 'top-left', x: 0.1, y: 0.1 },
      { id: 'top-center', x: 0.5, y: 0.1 },
      { id: 'top-right', x: 0.9, y: 0.1 },
      { id: 'center-left', x: 0.1, y: 0.5 },
      { id: 'center', x: 0.5, y: 0.5 },
      { id: 'center-right', x: 0.9, y: 0.5 },
      { id: 'bottom-left', x: 0.1, y: 0.9 },
      { id: 'bottom-center', x: 0.5, y: 0.9 },
      { id: 'bottom-right', x: 0.9, y: 0.9 },
    ],
    samplesPerPoint: 4,
    maxAttemptsPerPoint: 8,
  };
}

/**
 * Handles the gnosis classify Eye Tracking Zone workflow.
 */
export function classifyEyeTrackingZone(
  point: GazePoint | null
): EyeTrackingZone {
  if (point === null) {
    return 'unavailable';
  }

  const x = Math.max(0, Math.min(0.999999, point.x));
  const y = Math.max(0, Math.min(0.999999, point.y));
  const column = Math.min(2, Math.floor(x * 3));
  const row = Math.min(2, Math.floor(y * 3));

  return EYE_TRACKING_ZONES[row * 3 + column] ?? 'unavailable';
}

/**
 * Creates the Eye Tracking Session.
 */
export function createEyeTrackingSession(input: {
  sessionId?: string;
  startedAt?: number;
  status?: EyeTrackingSessionStatus;
  deviceId?: string;
  modelVersion?: string;
  calibrationPlan?: EyeTrackingCalibrationPlan;
  metadata?: Record<string, unknown>;
} = {}): EyeTrackingSession {
  const plan = cloneCalibrationPlan(
    input.calibrationPlan ?? defaultEyeTrackingCalibrationPlan()
  );

  return {
    sessionId: input.sessionId ?? crypto.randomUUID(),
    startedAt: input.startedAt ?? Date.now(),
    status: input.status ?? 'calibrating',
    ...(input.deviceId !== undefined ? { deviceId: input.deviceId } : {}),
    ...(input.modelVersion !== undefined
      ? { modelVersion: input.modelVersion }
      : {}),
    calibration: buildCalibrationLedger(plan, []),
    observations: [],
    calibrationSamples: [],
    failures: [],
    boundary: createVoidBoundary(EYE_TRACKING_FAILURE_DIMENSIONS.length),
    metadata: { ...(input.metadata ?? {}) },
  };
}

/**
 * Creates the Eye Tracking Corpus.
 */
export function createEyeTrackingCorpus(): EyeTrackingCorpus {
  return {
    sessions: [],
    boundary: createVoidBoundary(EYE_TRACKING_FAILURE_DIMENSIONS.length),
  };
}

/**
 * Handles the gnosis append Eye Tracking Observation workflow.
 */
export function appendEyeTrackingObservation(
  session: EyeTrackingSession,
  observation: Omit<EyeTrackingObservation, 'observationId' | 'zone'> & {
    observationId?: string;
    zone?: EyeTrackingZone;
  }
): EyeTrackingObservation {
  const canonical: EyeTrackingObservation = {
    observationId: observation.observationId ?? crypto.randomUUID(),
    timestamp: observation.timestamp,
    left: cloneEyeReading(observation.left),
    right: cloneEyeReading(observation.right),
    gazePoint:
      observation.gazePoint === null ? null : clonePoint(observation.gazePoint),
    zone: observation.zone ?? classifyEyeTrackingZone(observation.gazePoint),
    confidence: observation.confidence,
    blink: observation.blink,
    eyeBlocked: observation.eyeBlocked,
    fixationDurationMs: observation.fixationDurationMs,
    saccadeVelocity: observation.saccadeVelocity,
    scanningVariance: observation.scanningVariance,
    ...(observation.faceQuality !== undefined
      ? { faceQuality: observation.faceQuality }
      : {}),
    ...(observation.source !== undefined ? { source: observation.source } : {}),
    ...(observation.metadata !== undefined
      ? { metadata: { ...observation.metadata } }
      : {}),
  };

  session.observations.push(canonical);
  return canonical;
}

/**
 * Handles the gnosis record Eye Tracking Failure workflow.
 */
export function recordEyeTrackingFailure(
  session: EyeTrackingSession,
  failure: Omit<EyeTrackingFailureEvent, 'failureId' | 'sessionId' | 'magnitude'> & {
    failureId?: string;
    magnitude?: number;
  }
): EyeTrackingFailureEvent {
  const severity = failure.severity;
  const kind = failure.kind;
  const magnitude = normalizeMagnitude(failure.magnitude, severity);
  const dimension = FAILURE_KIND_TO_DIMENSION.get(kind);

  if (dimension === undefined) {
    throw new Error(`Unknown eye-tracking failure kind "${kind}".`);
  }

  const canonical: EyeTrackingFailureEvent = {
    failureId: failure.failureId ?? crypto.randomUUID(),
    sessionId: session.sessionId,
    timestamp: failure.timestamp,
    kind,
    magnitude,
    severity,
    message: failure.message,
    ...(failure.observationId !== undefined
      ? { observationId: failure.observationId }
      : {}),
    ...(failure.expectedZone !== undefined
      ? { expectedZone: failure.expectedZone }
      : {}),
    ...(failure.observedZone !== undefined
      ? { observedZone: failure.observedZone }
      : {}),
    ...(failure.metadata !== undefined
      ? { metadata: { ...failure.metadata } }
      : {}),
  };

  session.failures.push(canonical);
  updateVoidBoundary(session.boundary, dimension, magnitude);
  return canonical;
}

/**
 * Handles the gnosis record Eye Tracking Calibration Sample workflow.
 */
export function recordEyeTrackingCalibrationSample(
  session: EyeTrackingSession,
  sample: Omit<
    EyeTrackingCalibrationSample,
    'sampleId' | 'sessionId' | 'point' | 'outcome'
  > & {
    sampleId?: string;
    outcome?: 'accepted' | 'rejected';
  }
): EyeTrackingCalibrationSample {
  const point = session.calibration.plan.points.find(
    (candidate) => candidate.id === sample.pointId
  );
  const priorPointStatus = session.calibration.points.find(
    (candidate) => candidate.pointId === sample.pointId
  );

  let outcome: 'accepted' | 'rejected' = sample.outcome ?? 'accepted';
  let rejectionReason = sample.rejectionReason;

  if (!point) {
    outcome = 'rejected';
    rejectionReason = 'unknown-point';
  } else if (priorPointStatus?.complete) {
    outcome = 'rejected';
    rejectionReason = 'point-complete';
  } else if (priorPointStatus?.exhausted) {
    outcome = 'rejected';
    rejectionReason = 'point-exhausted';
  }

  const canonical: EyeTrackingCalibrationSample = {
    sampleId: sample.sampleId ?? crypto.randomUUID(),
    sessionId: session.sessionId,
    pointId: sample.pointId,
    point: point ? clonePoint(point) : { x: 0, y: 0 },
    observationId: sample.observationId,
    timestamp: sample.timestamp,
    quality: sample.quality,
    outcome,
    ...(rejectionReason !== undefined ? { rejectionReason } : {}),
  };

  session.calibrationSamples.push(canonical);
  recomputeCalibration(session);

  if (canonical.outcome === 'rejected' && isFailureKind(canonical.rejectionReason)) {
    recordEyeTrackingFailure(session, {
      timestamp: canonical.timestamp,
      kind: canonical.rejectionReason,
      severity: 'warning',
      message: `Calibration sample rejected: ${canonical.rejectionReason}`,
      observationId: canonical.observationId,
      metadata: {
        pointId: canonical.pointId,
        quality: canonical.quality,
      },
    });
  }

  return canonical;
}

/**
 * Handles the gnosis complete Eye Tracking Session workflow.
 */
export function completeEyeTrackingSession(
  session: EyeTrackingSession,
  status: Extract<EyeTrackingSessionStatus, 'completed' | 'failed'> = 'completed',
  endedAt: number = Date.now()
): EyeTrackingSession {
  session.status = status;
  session.endedAt = endedAt;
  return session;
}

/**
 * Handles the gnosis ingest Eye Tracking Session workflow.
 */
export function ingestEyeTrackingSession(
  corpus: EyeTrackingCorpus,
  session: EyeTrackingSession
): EyeTrackingSession {
  const canonical = cloneSession(session);
  corpus.sessions.push(canonical);

  for (let index = 0; index < canonical.boundary.counts.length; index++) {
    const magnitude = canonical.boundary.counts[index];
    if (typeof magnitude === 'number' && magnitude > 0) {
      updateVoidBoundary(corpus.boundary, index, magnitude);
    }
  }

  return canonical;
}

/**
 * Creates the Eye Tracking Vent Records.
 */
export function createEyeTrackingVentRecords(
  session: EyeTrackingSession
): VoidVentRecord[] {
  return session.failures.map((failure, index) => ({
    dimensionIdx: FAILURE_KIND_TO_DIMENSION.get(failure.kind) ?? 0,
    magnitude: failure.magnitude,
    sequenceNumber: index + 1,
    timestamp: failure.timestamp,
  }));
}

/**
 * Builds the Eye Tracking Merkle Tree.
 */
export function buildEyeTrackingMerkleTree(
  session: EyeTrackingSession
): VoidMerkleTree {
  return buildVoidMerkleTree(createEyeTrackingVentRecords(session));
}

/**
 * Handles the gnosis audit Eye Tracking Session workflow.
 */
export function auditEyeTrackingSession(
  session: EyeTrackingSession
): EyeTrackingAuditReport {
  const buleyean = fullBuleyeanAudit({
    rounds: session.boundary.totalEntries,
    rejections: [...session.boundary.counts],
  });
  const merkle = buildEyeTrackingMerkleTree(session);
  const invariants: EyeTrackingInvariant[] = [];

  const deficitMatches =
    session.calibration.deficit ===
    Math.max(0, session.calibration.requiredSamples - session.calibration.acceptedSamples);
  invariants.push({
    id: 'calibration-deficit-explicit',
    theoremId: 'law_cave',
    pass: deficitMatches,
    details: deficitMatches
      ? `Deficit ${session.calibration.deficit} is explicit.`
      : 'Calibration deficit does not match required - accepted.',
  });

  const conservationHolds =
    session.calibration.observedSamples ===
    session.calibration.acceptedSamples + session.calibration.rejectedSamples;
  invariants.push({
    id: 'calibration-conservation',
    theoremId: 'law_conservation',
    pass: conservationHolds,
    details: conservationHolds
      ? 'Observed calibration samples are conserved into accepted + rejected.'
      : 'Observed calibration samples are not conserved into accepted + rejected.',
  });

  const sharpZones = session.observations.every(
    (observation) =>
      EYE_TRACKING_ZONES.includes(observation.zone) &&
      (observation.zone === 'unavailable' || observation.gazePoint !== null)
  );
  invariants.push({
    id: 'zone-admission-sharp',
    theoremId: 'law_sorites_sharpness',
    pass: sharpZones,
    details: sharpZones
      ? 'Every observation resolves to exactly one admitted zone or unavailable.'
      : 'Some observations do not resolve to a sharp zone assignment.',
  });

  const finiteTermination =
    session.calibration.observedSamples <=
    session.calibration.plan.points.length *
      session.calibration.plan.maxAttemptsPerPoint;
  invariants.push({
    id: 'calibration-termination-finite',
    theoremId: 'law_termination',
    pass: finiteTermination,
    details: finiteTermination
      ? 'Calibration stays within the finite point/attempt schedule.'
      : 'Calibration exceeded the finite point/attempt schedule.',
  });

  const merkleMatches = merkle.size === session.failures.length;
  invariants.push({
    id: 'merkle-log-complete',
    theoremId: 'void_merkle_append_only',
    pass: merkleMatches,
    details: merkleMatches
      ? 'Merkle vent log matches failure history length.'
      : 'Merkle vent log size diverges from failure history length.',
  });

  return {
    pass: buleyean.pass && invariants.every((invariant) => invariant.pass),
    buleyean,
    invariants,
    merkle: {
      rootHash: merkle.rootHash,
      size: merkle.size,
    },
  };
}

/**
 * Converts input into Eye Tracking Rejection Record.
 */
export function toEyeTrackingRejectionRecord(
  session: EyeTrackingSession
): EyeTrackingRejectionRecord {
  const grouped = new Map<string, number>();

  for (const failure of session.failures) {
    const key = `${failure.kind}: ${failure.message}`;
    grouped.set(key, (grouped.get(key) ?? 0) + failure.magnitude);
  }

  const rejectedResponses = [...grouped.keys()];
  const rejectionCounts = rejectedResponses.map(
    (response) => grouped.get(response) ?? 0
  );
  const merkle = buildEyeTrackingMerkleTree(session);

  return {
    prompt: `eyetracking:${session.sessionId}`,
    rejectedResponses,
    rejectionCounts,
    totalRounds: session.boundary.totalEntries,
    metadata: {
      sessionId: session.sessionId,
      status: session.status,
      startedAt: session.startedAt,
      endedAt: session.endedAt ?? null,
      deviceId: session.deviceId ?? null,
      modelVersion: session.modelVersion ?? null,
      calibrationDeficit: session.calibration.deficit,
      observationCount: session.observations.length,
      failureCount: session.failures.length,
      merkleRoot: merkle.rootHash,
    },
  };
}

/**
 * Handles the gnosis serialize Eye Tracking Rejection JSONL workflow.
 */
export function serializeEyeTrackingRejectionJSONL(
  sessions: readonly EyeTrackingSession[]
): string {
  return sessions
    .map((session) => {
      const record = toEyeTrackingRejectionRecord(session);
      return JSON.stringify({
        prompt: record.prompt,
        rejected_responses: record.rejectedResponses,
        rejection_counts: record.rejectionCounts,
        total_rounds: record.totalRounds,
        metadata: record.metadata,
      });
    })
    .join('\n');
}

/**
 * Handles the gnosis summarize Eye Tracking Failure Pressure workflow.
 */
export function summarizeEyeTrackingFailurePressure(
  source:
    | EyeTrackingCorpus
    | EyeTrackingSession
    | readonly EyeTrackingSession[]
): EyeTrackingFailurePressure {
  const sessions = toEyeTrackingSessions(source);

  const byKind = createEmptyFailureCounts();
  const byDevice: Record<string, number> = {};
  const byModel: Record<string, number> = {};

  for (const session of sessions) {
    for (const failure of session.failures) {
      byKind[failure.kind] += failure.magnitude;

      if (session.deviceId) {
        byDevice[session.deviceId] =
          (byDevice[session.deviceId] ?? 0) + failure.magnitude;
      }

      if (session.modelVersion) {
        byModel[session.modelVersion] =
          (byModel[session.modelVersion] ?? 0) + failure.magnitude;
      }
    }
  }

  let hottestKind: EyeTrackingFailureKind | null = null;
  let hottestMagnitude = -1;
  for (const kind of EYE_TRACKING_FAILURE_DIMENSIONS) {
    if (byKind[kind] > hottestMagnitude) {
      hottestKind = kind;
      hottestMagnitude = byKind[kind];
    }
  }

  return {
    sessionCount: sessions.length,
    totalMagnitude: sessions.reduce(
      (sum, session) => sum + session.boundary.totalEntries,
      0
    ),
    byKind,
    byDevice,
    byModel,
    hottestKind,
  };
}

/**
 * Handles the gnosis select Eye Tracking Sessions workflow.
 */
export function selectEyeTrackingSessions(
  source:
    | EyeTrackingCorpus
    | EyeTrackingSession
    | readonly EyeTrackingSession[],
  selection: EyeTrackingDatasetSelection = {}
): EyeTrackingSession[] {
  const deviceIds =
    selection.deviceIds === undefined
      ? null
      : new Set(selection.deviceIds.map((value) => value.trim()));
  const modelVersions =
    selection.modelVersions === undefined
      ? null
      : new Set(selection.modelVersions.map((value) => value.trim()));
  const statuses =
    selection.statuses === undefined ? null : new Set(selection.statuses);
  const failureKinds =
    selection.failureKinds === undefined
      ? null
      : new Set(selection.failureKinds);

  const sessions = toEyeTrackingSessions(source)
    .filter((session) => {
      if (deviceIds && !deviceIds.has(session.deviceId ?? '')) {
        return false;
      }

      if (modelVersions && !modelVersions.has(session.modelVersion ?? '')) {
        return false;
      }

      if (statuses && !statuses.has(session.status)) {
        return false;
      }

      if (selection.requireReadyCalibration && !session.calibration.ready) {
        return false;
      }

      if (failureKinds) {
        const matchesFailureKind = session.failures.some((failure) =>
          failureKinds.has(failure.kind)
        );
        if (!matchesFailureKind) {
          return false;
        }
      }

      if (
        selection.minTotalFailureMagnitude !== undefined &&
        session.boundary.totalEntries < selection.minTotalFailureMagnitude
      ) {
        return false;
      }

      return true;
    })
    .map(cloneSession);

  sessions.sort((left, right) => {
    if (selection.sortBy === 'failureMagnitude') {
      const delta = right.boundary.totalEntries - left.boundary.totalEntries;
      if (delta !== 0) {
        return delta;
      }
    }

    return left.startedAt - right.startedAt;
  });

  if (
    selection.maxSessions !== undefined &&
    selection.maxSessions >= 0 &&
    selection.maxSessions < sessions.length
  ) {
    return sessions.slice(0, selection.maxSessions);
  }

  return sessions;
}

/**
 * Creates the Eye Tracking Dataset Slice.
 */
export function createEyeTrackingDatasetSlice(
  source:
    | EyeTrackingCorpus
    | EyeTrackingSession
    | readonly EyeTrackingSession[],
  options: {
    split?: EyeTrackingDatasetSplit;
    selection?: EyeTrackingDatasetSelection;
    createdAt?: number;
    metadata?: Record<string, unknown>;
  } = {}
): EyeTrackingDatasetSlice {
  const sessions = selectEyeTrackingSessions(source, options.selection);
  const split = options.split ?? 'train';
  const createdAt = options.createdAt ?? Date.now();
  const merkle = sessions.map((session) => {
    const tree = buildEyeTrackingMerkleTree(session);
    return {
      sessionId: session.sessionId,
      rootHash: tree.rootHash,
      size: tree.size,
    };
  });
  const boundary = createEyeTrackingBoundarySummary(sessions);
  const signature = createStableSessionSignature(sessions);

  return {
    sliceId: `eye-slice:${split}:${signature}`,
    split,
    createdAt,
    sessionIds: sessions.map((session) => session.sessionId),
    sessionCount: sessions.length,
    sessions,
    totalObservations: sessions.reduce(
      (sum, session) => sum + session.observations.length,
      0
    ),
    totalFailures: sessions.reduce(
      (sum, session) => sum + session.failures.length,
      0
    ),
    totalCalibrationDeficit: sessions.reduce(
      (sum, session) => sum + session.calibration.deficit,
      0
    ),
    boundary,
    pressure: summarizeEyeTrackingFailurePressure(sessions),
    merkle,
    rejectionJsonl: serializeEyeTrackingRejectionJSONL(sessions),
    metadata: { ...(options.metadata ?? {}) },
  };
}

/**
 * Creates the Eye Tracking Training Manifest.
 */
export function createEyeTrackingTrainingManifest(
  inputs: readonly EyeTrackingTrainingSplitInput[],
  options: {
    generatedAt?: number;
    metadata?: Record<string, unknown>;
  } = {}
): EyeTrackingTrainingManifest {
  const generatedAt = options.generatedAt ?? Date.now();
  const slices = inputs.map((input) => {
    const sliceOptions: {
      split: EyeTrackingDatasetSplit;
      createdAt: number;
      selection?: EyeTrackingDatasetSelection;
      metadata?: Record<string, unknown>;
    } = {
      split: input.split,
      createdAt: generatedAt,
    };
    if (input.selection !== undefined) {
      sliceOptions.selection = cloneSelection(input.selection);
    }
    if (input.metadata !== undefined) {
      sliceOptions.metadata = { ...input.metadata };
    }
    return createEyeTrackingDatasetSlice(input.source, sliceOptions);
  });
  const boundary = createEyeTrackingBoundarySummary(
    slices.flatMap((slice) => slice.sessions)
  );
  const pressure = summarizeEyeTrackingFailurePressure(
    slices.flatMap((slice) => slice.sessions)
  );
  const hottestKinds = EYE_TRACKING_FAILURE_DIMENSIONS.filter(
    (kind) => pressure.byKind[kind] > 0
  ).sort((left, right) => pressure.byKind[right] - pressure.byKind[left]);
  const signature = createStableSessionSignature(
    slices.flatMap((slice) => slice.sessions)
  );

  return {
    manifestId: `eye-manifest:${signature}`,
    service: 'eyetracking-as-a-service',
    generatedAt,
    slices,
    totalSessionCount: slices.reduce(
      (sum, slice) => sum + slice.sessionCount,
      0
    ),
    totalObservationCount: slices.reduce(
      (sum, slice) => sum + slice.totalObservations,
      0
    ),
    totalFailureCount: slices.reduce(
      (sum, slice) => sum + slice.totalFailures,
      0
    ),
    totalCalibrationDeficit: slices.reduce(
      (sum, slice) => sum + slice.totalCalibrationDeficit,
      0
    ),
    boundary,
    hottestKinds,
    metadata: { ...(options.metadata ?? {}) },
  };
}

/**
 * Handles the gnosis default Eye Tracking Buleyean Training Config workflow.
 */
export function defaultEyeTrackingBuleyeanTrainingConfig(
  overrides: EyeTrackingBuleyeanTrainingConfigOverrides = {}
): EyeTrackingBuleyeanTrainingConfig {
  return {
    alpha: overrides.alpha ?? 0.7,
    contrastWeight: overrides.contrastWeight ?? 0.3,
    sftWeight: overrides.sftWeight ?? 0,
    temperature: overrides.temperature ?? 1,
    minRejectionCount: overrides.minRejectionCount ?? 1,
    sparse: overrides.sparse ?? true,
    curriculum: overrides.curriculum ?? 'rejection_density',
    federated: {
      dimensions:
        overrides.federated?.dimensions ??
        EYE_TRACKING_FAILURE_DIMENSIONS.length,
      lr: overrides.federated?.lr ?? 0.001,
      minRejections: overrides.federated?.minRejections ?? 100,
      maxGradNorm: overrides.federated?.maxGradNorm ?? 1,
      sparsityThreshold: overrides.federated?.sparsityThreshold ?? 0.001,
      nodeId: overrides.federated?.nodeId ?? 'dash-eyetracking',
    },
  };
}

function materializeEyeTrackingBuleyeanInput(
  input: EyeTrackingTrainingSplitInput
): EyeTrackingBuleyeanTrainingJobInput {
  const materialized: EyeTrackingBuleyeanTrainingJobInput = {
    split: input.split,
    sessions: selectEyeTrackingSessions(input.source, input.selection).map(
      (session) => cloneSession(session)
    ),
  };
  if (input.selection !== undefined) {
    materialized.selection = cloneSelection(input.selection);
  }
  if (input.metadata !== undefined) {
    materialized.metadata = { ...input.metadata };
  }
  return materialized;
}

function createEyeTrackingBuleyeanSplitSummary(
  input: EyeTrackingBuleyeanTrainingJobInput
): EyeTrackingBuleyeanSplitSummary {
  const records = input.sessions.map((session) =>
    toEyeTrackingRejectionRecord(session)
  );

  return {
    split: input.split,
    sessionIds: input.sessions.map((session) => session.sessionId),
    sessionCount: input.sessions.length,
    recordCount: records.length,
    totalRounds: records.reduce((sum, record) => sum + record.totalRounds, 0),
    totalRejectionMass: records.reduce(
      (sum, record) =>
        sum +
        record.rejectionCounts.reduce(
          (recordSum, count) => recordSum + count,
          0
        ),
      0
    ),
  };
}

function createEyeTrackingBuleyeanTrainingJobId(
  inputs: readonly EyeTrackingBuleyeanTrainingJobInput[],
  config: EyeTrackingBuleyeanTrainingConfig
): string {
  const source = inputs
    .map((input) =>
      [
        input.split,
        createStableSessionSignature(input.sessions),
        stableStringify(input.selection ?? null),
      ].join(':')
    )
    .join('|');

  return `eye-buleyean-job:${createStableHash(
    `${source}|${stableStringify(config)}`
  )}`;
}

function createEyeTrackingBuleyeanDeltaId(
  sessions: readonly EyeTrackingSession[],
  selection: EyeTrackingDatasetSelection | undefined
): string {
  return `eye-buleyean-delta:${createStableHash(
    `${createStableSessionSignature(sessions)}|${stableStringify(
      selection ?? null
    )}`
  )}`;
}

function buildEyeTrackingBuleyeanTrainingJob(
  inputs: readonly EyeTrackingBuleyeanTrainingJobInput[],
  options: EyeTrackingBuleyeanTrainingJobOptions = {}
): EyeTrackingBuleyeanTrainingJob {
  const createdAt = options.createdAt ?? Date.now();
  const config = defaultEyeTrackingBuleyeanTrainingConfig(options.config);
  const manifestInputs = inputs.map((input) => {
    const manifestInput: EyeTrackingTrainingSplitInput = {
      split: input.split,
      source: input.sessions,
    };
    if (input.selection !== undefined) {
      manifestInput.selection = cloneSelection(input.selection);
    }
    if (input.metadata !== undefined) {
      manifestInput.metadata = { ...input.metadata };
    }
    return manifestInput;
  });
  const manifestOptions: {
    generatedAt: number;
    metadata?: Record<string, unknown>;
  } = {
    generatedAt: createdAt,
  };
  if (options.metadata !== undefined) {
    manifestOptions.metadata = options.metadata;
  }
  const manifest = createEyeTrackingTrainingManifest(
    manifestInputs,
    manifestOptions
  );
  const sessions = inputs.flatMap((input) => input.sessions);
  const records = sessions.map((session) => toEyeTrackingRejectionRecord(session));

  return {
    jobId: createEyeTrackingBuleyeanTrainingJobId(inputs, config),
    createdAt,
    service: 'eyetracking-as-a-service',
    trainer: 'buleyean-rl',
    inputs: inputs.map((input) =>
      cloneEyeTrackingBuleyeanTrainingJobInput(input)
    ),
    manifest,
    splits: inputs.map((input) => createEyeTrackingBuleyeanSplitSummary(input)),
    records,
    rejectionJsonl: serializeEyeTrackingRejectionJSONL(sessions),
    corpus: summarizeEyeTrackingBuleyeanTrainingCorpus(sessions),
    config,
    metadata: { ...(options.metadata ?? {}) },
  };
}

/**
 * Handles the gnosis summarize Eye Tracking Buleyean Training Corpus workflow.
 */
export function summarizeEyeTrackingBuleyeanTrainingCorpus(
  source:
    | EyeTrackingCorpus
    | EyeTrackingSession
    | readonly EyeTrackingSession[],
  options: EyeTrackingBuleyeanCorpusSummaryOptions = {}
): EyeTrackingBuleyeanCorpusSummary {
  const sessions = selectEyeTrackingSessions(source, options.selection);
  const pressure = summarizeEyeTrackingFailurePressure(sessions);
  const byStatus = createEmptyStatusCounts();

  let totalRejectedResponses = 0;
  let totalRejectionMass = 0;
  let totalRounds = 0;

  for (const session of sessions) {
    byStatus[session.status] += 1;

    const record = toEyeTrackingRejectionRecord(session);
    totalRejectedResponses += record.rejectedResponses.length;
    totalRejectionMass += record.rejectionCounts.reduce(
      (sum, count) => sum + count,
      0
    );
    totalRounds += record.totalRounds;
  }

  const promptCount = sessions.length;

  return {
    promptCount,
    sessionCount: sessions.length,
    totalRejectedResponses,
    totalRejectionMass,
    totalRounds,
    totalObservations: sessions.reduce(
      (sum, session) => sum + session.observations.length,
      0
    ),
    totalCalibrationDeficit: sessions.reduce(
      (sum, session) => sum + session.calibration.deficit,
      0
    ),
    averageRejectedResponsesPerPrompt:
      promptCount === 0 ? 0 : totalRejectedResponses / promptCount,
    averageRejectionMassPerPrompt:
      promptCount === 0 ? 0 : totalRejectionMass / promptCount,
    byKind: pressure.byKind,
    byDevice: pressure.byDevice,
    byModel: pressure.byModel,
    byStatus,
    hottestKind: pressure.hottestKind,
  };
}

/**
 * Creates the Eye Tracking Buleyean Training Job.
 */
export function createEyeTrackingBuleyeanTrainingJob(
  inputs: readonly EyeTrackingTrainingSplitInput[],
  options: EyeTrackingBuleyeanTrainingJobOptions = {}
): EyeTrackingBuleyeanTrainingJob {
  return buildEyeTrackingBuleyeanTrainingJob(
    inputs.map((input) => materializeEyeTrackingBuleyeanInput(input)),
    options
  );
}

/**
 * Handles the gnosis append Eye Tracking Buleyean Rejection Batch workflow.
 */
export function appendEyeTrackingBuleyeanRejectionBatch(
  job: EyeTrackingBuleyeanTrainingJob,
  input: EyeTrackingTrainingSplitInput,
  options: EyeTrackingBuleyeanAppendOptions = {}
): EyeTrackingBuleyeanAppendResult {
  const materialized = materializeEyeTrackingBuleyeanInput(input);
  const seenSessionIds = new Set(
    job.inputs.flatMap((entry) =>
      entry.sessions.map((session) => session.sessionId)
    )
  );
  const appendedSessionIds: string[] = [];
  const duplicateSessionIds: string[] = [];
  const dedupedSessions: EyeTrackingSession[] = [];

  for (const session of materialized.sessions) {
    if (seenSessionIds.has(session.sessionId)) {
      duplicateSessionIds.push(session.sessionId);
      continue;
    }

    seenSessionIds.add(session.sessionId);
    appendedSessionIds.push(session.sessionId);
    dedupedSessions.push(session);
  }

  if (dedupedSessions.length === 0) {
    return {
      job,
      appendedSessionIds,
      duplicateSessionIds,
    };
  }

  const updatedInputs: EyeTrackingBuleyeanTrainingJobInput[] = [
    ...job.inputs.map((entry) => cloneEyeTrackingBuleyeanTrainingJobInput(entry)),
    cloneEyeTrackingBuleyeanTrainingJobInput(materialized),
  ];
  const appendOptions: EyeTrackingBuleyeanTrainingJobOptions = {};
  if (options.createdAt !== undefined) {
    appendOptions.createdAt = options.createdAt;
  }
  appendOptions.config = job.config;
  appendOptions.metadata = {
    ...job.metadata,
    ...(options.metadata ?? {}),
    parentJobId: job.jobId,
  };

  return {
    job: buildEyeTrackingBuleyeanTrainingJob(updatedInputs, appendOptions),
    appendedSessionIds,
    duplicateSessionIds,
  };
}

/**
 * Handles the gnosis export Eye Tracking Buleyean Personalization Delta workflow.
 */
export function exportEyeTrackingBuleyeanPersonalizationDelta(
  source:
    | EyeTrackingCorpus
    | EyeTrackingSession
    | readonly EyeTrackingSession[],
  options: EyeTrackingBuleyeanPersonalizationDeltaOptions = {}
): EyeTrackingBuleyeanPersonalizationDelta {
  const sessions = selectEyeTrackingSessions(source, options.selection);
  const selection =
    options.selection !== undefined
      ? cloneSelection(options.selection)
      : undefined;

  return {
    deltaId: createEyeTrackingBuleyeanDeltaId(sessions, selection),
    createdAt: options.createdAt ?? Date.now(),
    trainer: 'buleyean-rl',
    sessionIds: sessions.map((session) => session.sessionId),
    sessionCount: sessions.length,
    records: sessions.map((session) => toEyeTrackingRejectionRecord(session)),
    rejectionJsonl: serializeEyeTrackingRejectionJSONL(sessions),
    corpus: summarizeEyeTrackingBuleyeanTrainingCorpus(sessions),
    ...(selection !== undefined ? { selection } : {}),
    metadata: { ...(options.metadata ?? {}) },
  };
}

/**
 * Handles the gnosis encode Eye Tracking Failure Event workflow.
 */
export function encodeEyeTrackingFailureEvent(
  failure: EyeTrackingFailureEvent
): Uint8Array {
  return new TextEncoder().encode(JSON.stringify(failure));
}

/**
 * Handles the gnosis decode Eye Tracking Failure Event workflow.
 */
export function decodeEyeTrackingFailureEvent(
  payload: Uint8Array
): EyeTrackingFailureEvent {
  const decoded = JSON.parse(new TextDecoder().decode(payload)) as EyeTrackingFailureEvent;
  return cloneFailure(decoded);
}

/**
 * Creates the Eye Tracking QDoc.
 */
export function createEyeTrackingQDoc(guid?: string): QDoc {
  return new QDoc({ guid: guid ?? `eyetracking:${crypto.randomUUID()}` });
}

/**
 * Handles the gnosis store Eye Tracking Session In QDoc workflow.
 */
export function storeEyeTrackingSessionInQDoc(
  doc: QDoc,
  session: EyeTrackingSession,
  aliases: readonly string[] = []
): EyeTrackingSession {
  const canonical = cloneSession(session);
  const sessions = doc.getMap<EyeTrackingSession>(SESSION_MAP);
  const aliasMap = doc.getMap<string>(SESSION_ALIAS_MAP);

  sessions.set(canonical.sessionId, canonical);
  aliasMap.set(CURRENT_SESSION_ALIAS, canonical.sessionId);

  for (const alias of aliases
    .map(normalizeAlias)
    .filter((value) => value.length > 0)
    .sort()) {
    aliasMap.set(alias, canonical.sessionId);
  }

  return canonical;
}

/**
 * Loads the Eye Tracking Session From QDoc.
 */
export function loadEyeTrackingSessionFromQDoc(
  doc: QDoc,
  sessionId: string
): EyeTrackingSession | null {
  const normalizedSessionId = normalizeAlias(sessionId);
  if (normalizedSessionId.length === 0) {
    return null;
  }

  const session =
    doc.getMap<EyeTrackingSession>(SESSION_MAP).get(normalizedSessionId) ?? null;
  return session === null ? null : cloneSession(session);
}

/**
 * Loads the Eye Tracking Session By Alias From QDoc.
 */
export function loadEyeTrackingSessionByAliasFromQDoc(
  doc: QDoc,
  alias = CURRENT_SESSION_ALIAS
): EyeTrackingSession | null {
  const normalizedAlias = normalizeAlias(alias);
  if (normalizedAlias.length === 0) {
    return null;
  }

  const mappedSessionId =
    doc.getMap<string>(SESSION_ALIAS_MAP).get(normalizedAlias) ?? normalizedAlias;
  return loadEyeTrackingSessionFromQDoc(doc, mappedSessionId);
}

/**
 * Handles the gnosis append Eye Tracking Observation To QDoc workflow.
 */
export function appendEyeTrackingObservationToQDoc(
  doc: QDoc,
  sessionLookup: string,
  observation: Omit<EyeTrackingObservation, 'observationId' | 'zone'> & {
    observationId?: string;
    zone?: EyeTrackingZone;
  }
): EyeTrackingSession {
  const session = loadEyeTrackingSessionByAliasFromQDoc(doc, sessionLookup);
  if (!session) {
    throw new Error(`No eye-tracking session found for "${sessionLookup}".`);
  }

  appendEyeTrackingObservation(session, observation);
  return storeEyeTrackingSessionInQDoc(doc, session, [sessionLookup]);
}

/**
 * Handles the gnosis append Eye Tracking Failure To QDoc workflow.
 */
export function appendEyeTrackingFailureToQDoc(
  doc: QDoc,
  sessionLookup: string,
  failure: Omit<EyeTrackingFailureEvent, 'failureId' | 'sessionId' | 'magnitude'> & {
    failureId?: string;
    magnitude?: number;
  }
): EyeTrackingSession {
  const session = loadEyeTrackingSessionByAliasFromQDoc(doc, sessionLookup);
  if (!session) {
    throw new Error(`No eye-tracking session found for "${sessionLookup}".`);
  }

  recordEyeTrackingFailure(session, failure);
  return storeEyeTrackingSessionInQDoc(doc, session, [sessionLookup]);
}
