// @a0n/.ignored_gnosis/src/foil-edge-gate.ts
import {
  Verdict,
  isAccept,
  isReject,
  isDefer,
} from '@a0n/bitwise/triton-verdict';

import {
  evaluateTritonAdmission,
  type TritonAdmissionWork,
} from './gnosis-time/triton-admission-gate';

export { Verdict, isAccept, isReject, isDefer };

export const FOIL_EDGE_PROOF_FORMAT = 'gnosis-foil-zk-v1' as const;
export const FOIL_STATIC_PROVER_ID = 'lean4-static-prover' as const;

export type FoilEdgeTarget =
  | 'wasm-verifier'
  | 'webgpu-pipeline'
  | 'bare-metal-controller';

export interface ForkRaceFoldCertificateSummary {
  readonly classicalDeficit: number;
  readonly entropyEliminated: boolean;
  readonly voidBound: number;
  readonly edgeCount: number;
}

export interface FoilAdmissionProof {
  readonly format: typeof FOIL_EDGE_PROOF_FORMAT;
  readonly verifiedBy: typeof FOIL_STATIC_PROVER_ID;
  readonly proofBytes: number;
  readonly verificationCost: number;
  readonly constantBound: number;
  readonly certificate: ForkRaceFoldCertificateSummary;
}

export interface TopologyPayload {
  readonly ggGraph: ArrayBuffer | Uint8Array;
  readonly zkProof: FoilAdmissionProof;
  readonly tritonAdmission?: TritonAdmissionWork;
}

export interface FoilAdmissionResult {
  /**
   * Lossy collapse of {@link verdict}: `isAccept(verdict)`. Back-compat field —
   * `true` exactly when the proof cleanly admits. NOTE: a deferring `Abstain`
   * (latent / recheck-later) collapses to `false` here, masquerading as a
   * Decline, because a bool has no neutral middle. Read `verdict` to recover it.
   */
  readonly accepted: boolean;
  readonly reason: string;
  /**
   * The canonical ternary admission decision — the value-add over the old bool.
   *   - Accept  → valid proof, clean admit
   *   - Decline → invalid/forged proof, hard block (fail closed)
   *   - Abstain → proof not-yet-verifiable / latent / recheck-later (defer)
   * Additive and OPTIONAL so pre-existing consumers that build result-shaped
   * literals (e.g. the EHLA harvested/latent path) keep compiling; every result
   * minted by this gate (`foilResult`/`verifyFoilAdmissionProof`) always sets it,
   * so it is present at runtime on the canonical admission path.
   */
  readonly verdict?: Verdict;
}

/** Build a {@link FoilAdmissionResult}, deriving the lossy `accepted` bool from the verdict. */
function foilResult(verdict: Verdict, reason: string): FoilAdmissionResult {
  return { accepted: isAccept(verdict), reason, verdict };
}

export interface FoilEdgeRuntimeOptions {
  readonly target?: FoilEdgeTarget;
  readonly verifier?: (proof: FoilAdmissionProof) => FoilAdmissionResult;
}

export interface FoilEdgeAdmissionMetrics {
  readonly admissionAttempts: number;
  readonly tritonPreflightAttempts: number;
  readonly tritonPreflightAdmitted: number;
  readonly tritonPreflightDropped: number;
  readonly verifierCalls: number;
  readonly verifierRejected: number;
  readonly executedTopologies: number;
}

export class FoilAdmissionRejectedError extends Error {
  /** The canonical verdict that produced this throw. Always a non-Accept verdict. */
  readonly verdict: Verdict;
  constructor(reason: string, verdict: Verdict = Verdict.Decline) {
    super(`Admission Rejected: ${reason}`);
    this.name = 'FoilAdmissionRejectedError';
    this.verdict = verdict;
  }
}

/**
 * Thrown when admission is DEFERRED (Abstain): the proof is not a forgery but is
 * not-yet-verifiable / latent / recheck-later. Subclasses
 * {@link FoilAdmissionRejectedError} so existing `instanceof` callers (which
 * treat any non-Accept as "not admitted") keep working unchanged, while new
 * callers can distinguish defer from hard reject via `instanceof` / `verdict`.
 */
export class FoilAdmissionDeferredError extends FoilAdmissionRejectedError {
  constructor(reason: string) {
    super(reason, Verdict.Abstain);
    this.name = 'FoilAdmissionDeferredError';
  }
}

function isNonNegativeInteger(value: number): boolean {
  return Number.isInteger(value) && value >= 0;
}

/**
 * Handles the gnosis verify Foil Admission Proof workflow.
 */
export function verifyFoilAdmissionProof(
  proof: FoilAdmissionProof
): FoilAdmissionResult {
  // Hard forgeries / structural invalidity → Decline (fail closed).
  if (proof.format !== FOIL_EDGE_PROOF_FORMAT) {
    return foilResult(Verdict.Decline, 'unsupported-proof-format');
  }
  if (proof.verifiedBy !== FOIL_STATIC_PROVER_ID) {
    return foilResult(Verdict.Decline, 'missing-static-lean-prover');
  }
  if (!Number.isInteger(proof.proofBytes) || proof.proofBytes <= 0) {
    return foilResult(Verdict.Decline, 'empty-zk-proof');
  }
  // Malformed cost bounds (non-integer / negative) are forgeries → Decline.
  if (
    !isNonNegativeInteger(proof.verificationCost) ||
    !isNonNegativeInteger(proof.constantBound)
  ) {
    return foilResult(Verdict.Decline, 'verifier-cost-bound-malformed');
  }
  // Cost over the constant bound is NOT a forgery: the proof is structurally
  // well-formed but its verification has not yet settled under the constant
  // bound (latent / recheck-later). This is the case the old bool squashed into
  // a hard reject — the value-add. Abstain (defer): re-check later.
  if (proof.verificationCost > proof.constantBound) {
    return foilResult(Verdict.Abstain, 'verifier-cost-exceeds-constant-bound');
  }
  if (proof.certificate.classicalDeficit !== 0) {
    return foilResult(Verdict.Decline, 'classical-deficit-nonzero');
  }
  if (!proof.certificate.entropyEliminated) {
    return foilResult(Verdict.Decline, 'branch-entropy-not-eliminated');
  }
  if (
    !isNonNegativeInteger(proof.certificate.voidBound) ||
    !isNonNegativeInteger(proof.certificate.edgeCount) ||
    proof.certificate.edgeCount > proof.certificate.voidBound
  ) {
    return foilResult(Verdict.Decline, 'void-bound-violated');
  }
  return foilResult(Verdict.Accept, 'foil-edge-admission-accepted');
}

/**
 * Handles the gnosis assert Foil Edge Admission workflow.
 */
export function assertFoilEdgeAdmission(
  proof: FoilAdmissionProof,
  verifier: (proof: FoilAdmissionProof) => FoilAdmissionResult =
    verifyFoilAdmissionProof
): FoilAdmissionResult {
  const result = verifier(proof);
  // Recover the verdict; for a verifier that only set the bool (verdict omitted),
  // fall back to the lossy `accepted` collapse: true→Accept, false→Decline. This
  // preserves the old `!accepted` throw contract exactly for legacy verifiers.
  const verdict =
    result.verdict ?? (result.accepted ? Verdict.Accept : Verdict.Decline);
  if (isAccept(verdict)) {
    return result;
  }
  // Non-Accept fails closed (preserving the old `!accepted` throw contract), but
  // the verdict picks the error: Abstain (defer/recheck-later) throws the more
  // specific FoilAdmissionDeferredError; Decline (forgery) throws the hard
  // FoilAdmissionRejectedError. Both are `instanceof FoilAdmissionRejectedError`.
  if (isDefer(verdict)) {
    throw new FoilAdmissionDeferredError(
      `Latent ForkRaceFoldCertificate. Deferred at Void Boundary (${result.reason}).`
    );
  }
  throw new FoilAdmissionRejectedError(
    `Invalid ForkRaceFoldCertificate. Dropped at Void Boundary (${result.reason}).`,
    verdict
  );
}

/**
 * Creates the Static Foil Admission Proof.
 */
export function createStaticFoilAdmissionProof(
  certificate: ForkRaceFoldCertificateSummary,
  options: {
    readonly proofBytes?: number;
    readonly verificationCost?: number;
    readonly constantBound?: number;
  } = {}
): FoilAdmissionProof {
  return {
    format: FOIL_EDGE_PROOF_FORMAT,
    verifiedBy: FOIL_STATIC_PROVER_ID,
    proofBytes: options.proofBytes ?? 1,
    verificationCost: options.verificationCost ?? 1,
    constantBound: options.constantBound ?? 1,
    certificate,
  };
}

export const ZERO_DEFICIT_FOIL_CERTIFICATE: ForkRaceFoldCertificateSummary = {
  classicalDeficit: 0,
  entropyEliminated: true,
  voidBound: 0,
  edgeCount: 0,
};

export class FoilEdgeRuntime {
  private readonly verifier: (proof: FoilAdmissionProof) => FoilAdmissionResult;
  private admissionAttempts = 0;
  private tritonPreflightAttempts = 0;
  private tritonPreflightAdmitted = 0;
  private tritonPreflightDropped = 0;
  private verifierCalls = 0;
  private verifierRejected = 0;
  private executedTopologies = 0;
  readonly target: FoilEdgeTarget;

  constructor(options: FoilEdgeRuntimeOptions = {}) {
    this.target = options.target ?? 'webgpu-pipeline';
    this.verifier = options.verifier ?? verifyFoilAdmissionProof;
  }

  admit(payload: TopologyPayload): FoilAdmissionResult {
    this.admissionAttempts += 1;
    if (payload.tritonAdmission !== undefined) {
      this.tritonPreflightAttempts += 1;
      const tritonResult = evaluateTritonAdmission(payload.tritonAdmission);
      if (tritonResult.status !== 'admit') {
        this.tritonPreflightDropped += 1;
        throw new FoilAdmissionRejectedError(
          `Invalid TritonAdmission. Dropped at GnosisClock Boundary (${tritonResult.reason}).`
        );
      }
      this.tritonPreflightAdmitted += 1;
    }
    this.verifierCalls += 1;
    try {
      return assertFoilEdgeAdmission(payload.zkProof, this.verifier);
    } catch (error) {
      if (error instanceof FoilAdmissionRejectedError) {
        this.verifierRejected += 1;
      }
      throw error;
    }
  }

  executeTopology<T>(
    payload: TopologyPayload,
    executeAdmittedTopology: () => T
  ): T {
    this.admit(payload);
    const result = executeAdmittedTopology();
    this.executedTopologies += 1;
    return result;
  }

  metrics(): FoilEdgeAdmissionMetrics {
    return {
      admissionAttempts: this.admissionAttempts,
      tritonPreflightAttempts: this.tritonPreflightAttempts,
      tritonPreflightAdmitted: this.tritonPreflightAdmitted,
      tritonPreflightDropped: this.tritonPreflightDropped,
      verifierCalls: this.verifierCalls,
      verifierRejected: this.verifierRejected,
      executedTopologies: this.executedTopologies,
    };
  }
}
