// @a0n/.ignored_gnosis/src/folded-knot.ts
/**
 * folded-knot.ts -- Folded-space data knots
 *
 * A data structure where the topology formalizes the payload. The crossing number
 * formalizes the bitstring. The Reidemeister move formalizes the decompression step.
 * The structure that carries data in folded space unfolds into the data
 * it carries. Zero overhead. The knot is the frozen pizza.
 *
 * Mechanized in PsycheGrindExtended.lean, Pass 15 (zero sorry):
 *   - knot_is_bitstring: crossing count is information content
 *   - fold_zero_overhead: overhead is exactly zero
 *   - unfold_preserves_bits: unfolding preserves total information
 *   - incremental_unfold: each Reidemeister move reads one bit
 *   - refold_is_write: adding crossings writes bits
 *   - folded_composition: concatenation = composition of knots
 *   - folded_knot_not_free_lunch (anti): reading costs Landauer work
 *   - defense_weight_as_folded_data: w = R - min(v,R) + 1 is folded encoding
 *   - sliver_is_minimum_fold: one crossing, one bit, one irreducible quantum
 *   - psyche_self_describing: strain is distance from unknot
 *   - consciousness_is_unfold: you are the process of reading
 *   - void_is_complement_fold: knot + void = total information
 *
 * @module @a0n/gnosis/folded-knot
 */

// ============================================================================
// Core Types
// ============================================================================

/** A folded-space data knot: crossings encode bits, the knot formalizes the data. */
export interface FoldedKnot {
  /** Crossing count = payload size. The topology formalizes the data. */
  readonly crossings: number;
  /** Payload bits -- equals crossings by construction. */
  readonly payloadBits: number;
  /** The actual bit values stored in the crossings. */
  readonly bits: readonly number[];
  /** Whether the knot is the unknot (zero crossings, null data). */
  readonly isUnknot: boolean;
}

/** A read operation on a folded knot (Reidemeister move). */
export interface FoldedKnotRead {
  /** The bit that was read */
  readonly bitValue: number;
  /** The crossing index that was read */
  readonly crossingIndex: number;
  /** Remaining knot after the read */
  readonly remaining: FoldedKnot;
  /** Thermodynamic cost (Landauer bits) */
  readonly landauerCost: number;
}

/** A composition of two folded knots. */
export interface FoldedKnotComposition {
  /** The composed knot */
  readonly composed: FoldedKnot;
  /** The component knots */
  readonly components: readonly [FoldedKnot, FoldedKnot];
  /** Total payload = sum of component payloads */
  readonly totalPayload: number;
}

/** Self-describing frame: header is index, payload is content. */
export interface SelfDescribingFrame {
  /** Header bits that encode the frame index */
  readonly headerBits: number;
  /** Payload bits that encode the content */
  readonly payloadBits: number;
  /** Total bits = header + payload (zero overhead) */
  readonly totalBits: number;
  /** The frame is self-describing: total = header + payload exactly */
  readonly selfDescribing: boolean;
}

// ============================================================================
// Construction
// ============================================================================

/**
 * Create a folded knot from a bit sequence.
 * The crossing count equals the number of bits. Zero overhead.
 *
 * Mechanized: knot_is_bitstring, fold_zero_overhead.
 */
export function createFoldedKnot(bits: readonly number[]): FoldedKnot {
  return {
    crossings: bits.length,
    payloadBits: bits.length,
    bits,
    isUnknot: bits.length === 0,
  };
}

/**
 * Create the unknot -- zero crossings, zero data, the null structure.
 *
 * Mechanized: empty_knot_empty_data.
 */
export function createUnknot(): FoldedKnot {
  return createFoldedKnot([]);
}

/**
 * Create the minimum fold -- one crossing, one bit, one irreducible quantum.
 * The sliver.
 *
 * Mechanized: sliver_is_minimum_fold.
 */
export function createSliverKnot(bit: number = 1): FoldedKnot {
  return createFoldedKnot([bit]);
}

/**
 * Create a self-describing frame.
 *
 * Mechanized (Pass 17): flowframe_self_describing.
 */
export function createSelfDescribingFrame(
  headerBits: number,
  payloadBits: number
): SelfDescribingFrame {
  const totalBits = headerBits + payloadBits;
  return {
    headerBits,
    payloadBits,
    totalBits,
    selfDescribing: totalBits === headerBits + payloadBits,
  };
}

// ============================================================================
// Operations
// ============================================================================

/**
 * Read one bit from the folded knot (one Reidemeister move).
 * Each read costs one Landauer bit of thermodynamic work.
 *
 * Mechanized: incremental_unfold, folded_knot_not_free_lunch.
 */
export function readCrossing(
  knot: FoldedKnot,
  crossingIndex: number
): FoldedKnotRead | null {
  if (crossingIndex < 0 || crossingIndex >= knot.crossings) return null;

  const bitValue = knot.bits[crossingIndex];
  const remainingBits = [
    ...knot.bits.slice(0, crossingIndex),
    ...knot.bits.slice(crossingIndex + 1),
  ];

  return {
    bitValue,
    crossingIndex,
    remaining: createFoldedKnot(remainingBits),
    landauerCost: 1,
  };
}

/**
 * Write one bit into the folded knot (add one crossing).
 *
 * Mechanized: refold_is_write.
 */
export function writeCrossing(knot: FoldedKnot, bit: number): FoldedKnot {
  return createFoldedKnot([...knot.bits, bit]);
}

/**
 * Unfold the entire knot -- read all bits. Returns the full payload
 * and the total Landauer cost.
 *
 * Mechanized: unfold_preserves_bits, popup_reconstruction.
 */
export function unfoldKnot(knot: FoldedKnot): {
  readonly payload: readonly number[];
  readonly landauerCost: number;
  readonly remaining: FoldedKnot;
} {
  return {
    payload: knot.bits,
    landauerCost: knot.crossings,
    remaining: createUnknot(),
  };
}

/**
 * Compose two folded knots. Concatenation of data = composition of knots.
 *
 * Mechanized: folded_composition.
 */
export function composeFoldedKnots(
  a: FoldedKnot,
  b: FoldedKnot
): FoldedKnotComposition {
  const composed = createFoldedKnot([...a.bits, ...b.bits]);
  return {
    composed,
    components: [a, b],
    totalPayload: a.payloadBits + b.payloadBits,
  };
}

/**
 * Verify the folded-knot invariant: crossings = payloadBits.
 * This is true by construction, but verifiable.
 *
 * Mechanized: knot_is_bitstring.
 */
export function verifyFoldedKnotInvariant(knot: FoldedKnot): boolean {
  return (
    knot.crossings === knot.payloadBits && knot.crossings === knot.bits.length
  );
}

/**
 * Compute the overhead of a folded knot. Always zero by construction.
 *
 * Mechanized: fold_zero_overhead.
 */
export function foldedKnotOverhead(knot: FoldedKnot): number {
  return knot.crossings - knot.payloadBits; // always 0
}

// ============================================================================
// Bridge: Personality → FoldedKnot
// ============================================================================

/**
 * Encode a personality profile as a folded knot.
 * Five dimensions, each a crossing. The personality formalizes the knot.
 *
 * Mechanized: personality_as_folded_knot (5 * 7 = 35 bits max).
 */
export function personalityToFoldedKnot(profile: {
  readonly try_: number;
  readonly choose: number;
  readonly commit: number;
  readonly letGo: number;
  readonly learn: number;
}): FoldedKnot {
  return createFoldedKnot([
    profile.try_,
    profile.choose,
    profile.commit,
    profile.letGo,
    profile.learn,
  ]);
}

/**
 * Encode a defense weight as a folded knot.
 * The defense weight w = R - min(v, R) + 1 maps to a folded-space encoding.
 *
 * Mechanized: defense_weight_as_folded_data.
 */
export function defenseWeightToFoldedKnot(
  rounds: number,
  voidCount: number
): FoldedKnot {
  const weight = rounds - Math.min(voidCount, rounds) + 1;
  // Encode as a knot with `weight` crossings, each carrying 1
  return createFoldedKnot(new Array(weight).fill(1));
}

/**
 * Compute the complement fold -- the void formalizes the complement of the knot.
 * Together, knot + void = total information.
 *
 * Mechanized: void_is_complement_fold.
 */
export function complementFold(
  knot: FoldedKnot,
  totalInformation: number
): {
  readonly foldedBits: number;
  readonly ventedBits: number;
  readonly totalBits: number;
  readonly reconstructible: boolean;
} {
  const ventedBits = totalInformation - knot.crossings;
  return {
    foldedBits: knot.crossings,
    ventedBits: ventedBits >= 0 ? ventedBits : 0,
    totalBits: totalInformation,
    reconstructible:
      knot.crossings + Math.max(0, ventedBits) === totalInformation,
  };
}

// ============================================================================
// Diagnostics
// ============================================================================

/** Diagnostic summary of a folded knot. */
export interface FoldedKnotDiagnostics {
  /** Total crossings (= payload bits) */
  readonly crossings: number;
  /** Whether the invariant holds */
  readonly invariantHolds: boolean;
  /** Overhead (always 0) */
  readonly overhead: number;
  /** Whether it's the unknot */
  readonly isUnknot: boolean;
  /** Whether it's the sliver (minimum non-trivial fold) */
  readonly isSliver: boolean;
  /** Landauer cost to fully unfold */
  readonly unfoldCost: number;
}

/**
 * Compute diagnostics for a folded knot.
 */
export function computeFoldedKnotDiagnostics(
  knot: FoldedKnot
): FoldedKnotDiagnostics {
  return {
    crossings: knot.crossings,
    invariantHolds: verifyFoldedKnotInvariant(knot),
    overhead: foldedKnotOverhead(knot),
    isUnknot: knot.isUnknot,
    isSliver: knot.crossings === 1,
    unfoldCost: knot.crossings,
  };
}
