// @a0n/.ignored_gnosis/src/gg-test-harness.ts
'use aeon';

/**
 * GG Test Harness — Formal verification for .gg/.mgg topologies
 *
 * Wraps `checkGgProgram()` from `@a0n/aeon-logic` with
 * ergonomic assertion helpers. The topology is its own specification.
 * The model checker formalizes the linter.
 *
 * Usage:
 *   import { ggTest } from '@a0n/gnosis';
 *
 *   test('oscillator has bounded superposition', async () => {
 *     const result = await ggTest(oscillatorSource)
 *       .beta1Bounded(10)
 *       .safe()
 *       .eventuallyTerminates()
 *       .run();
 *
 *     expect(result.ok).toBe(true);
 *   });
 */

import {
  checkGgProgram,
  parseGgProgram,
  getGgRootNodeIds,
  getGgTerminalNodeIds,
  buildGgTemporalModel,
} from '@a0n/aeon-logic';
import type {
  GgTemporalModelOptions,
  GgTopologyState,
  CheckerOptions,
  CheckerResult,
  NamedPredicate,
} from '@a0n/aeon-logic';
import { lowerUfcsSource } from './ufcs';

// ═══════════════════════════════════════════════════════════════════════════════
// Test Result
// ═══════════════════════════════════════════════════════════════════════════════

export interface GGTestResult {
  /** Did all checks pass? */
  ok: boolean;
  /** Was the state space fully explored? */
  complete: boolean;
  /** Number of states explored */
  stateCount: number;
  /** Topology stats from the checker */
  topology: {
    forkCount: number;
    foldCount: number;
    ventCount: number;
    beta1: number;
    depthLayers: number;
  };
  /** Parsed topology info */
  program: {
    nodeCount: number;
    edgeCount: number;
    rootNodes: readonly string[];
    terminalNodes: readonly string[];
  };
  /** Violations (empty if ok) */
  violations: Array<{
    kind: 'invariant' | 'eventual';
    name: string;
    message: string;
  }>;
  /** Raw checker result */
  raw: CheckerResult<GgTopologyState>;
}

// ═══════════════════════════════════════════════════════════════════════════════
// Test Builder — Fluent API for constructing GG verification tests
// ═══════════════════════════════════════════════════════════════════════════════

export class GGTestBuilder {
  private source: string;
  private modelOpts: GgTemporalModelOptions = {};
  private extraInvariants: NamedPredicate<GgTopologyState>[] = [];
  private extraEventuals: NamedPredicate<GgTopologyState>[] = [];
  private _maxDepth = 64;
  private _maxBeta1 = 128;
  private _wantTerminates = false;
  private _wantCollapses = false;
  private _reachableNodes: string[] = [];
  private _reachableBeta1: number[] = [];

  constructor(source: string) {
    this.source = source;
  }

  /** Set the initial node for exploration */
  startAt(nodeId: string): this {
    this.modelOpts = { ...this.modelOpts, initialNodeId: nodeId };
    return this;
  }

  /** Set initial beta1 value */
  initialBeta1(value: number): this {
    this.modelOpts = { ...this.modelOpts, initialBeta1: value };
    return this;
  }

  /** Set maximum exploration depth */
  maxDepth(depth: number): this {
    this._maxDepth = depth;
    return this;
  }

  /** Assert beta1 stays below a bound (default invariant) */
  beta1Bounded(max: number): this {
    this._maxBeta1 = max;
    return this;
  }

  /** Assert the topology is safe: beta1 >= 0 and beta1 < bound */
  safe(): this {
    // These are the default invariants from aeon-logic.
    // Calling safe() is declarative — it says "I expect this to hold".
    return this;
  }

  /** Assert that all paths eventually reach a terminal node */
  eventuallyTerminates(): this {
    this._wantTerminates = true;
    return this;
  }

  /** Assert that all paths eventually collapse to beta1 = 0 */
  eventuallyCollapses(): this {
    this._wantCollapses = true;
    return this;
  }

  /** Add a custom invariant (must hold in every reachable state) */
  invariant(name: string, test: (state: GgTopologyState) => boolean): this {
    this.extraInvariants.push({ name, test });
    return this;
  }

  /** Add a custom eventual property (must hold in some reachable state on every path) */
  eventually(name: string, test: (state: GgTopologyState) => boolean): this {
    this.extraEventuals.push({ name, test });
    return this;
  }

  /** Assert that a specific node is reachable (exists in some explored state) */
  reachable(nodeId: string): this {
    this._reachableNodes.push(nodeId);
    return this;
  }

  /** Assert that beta1 reaches exactly a specific value at some point */
  reachesBeta1(value: number): this {
    this._reachableBeta1.push(value);
    return this;
  }

  /** Run the verification and return a structured result */
  async run(): Promise<GGTestResult> {
    const normalizedSource = lowerUfcsSource(this.source);
    const program = parseGgProgram(normalizedSource);
    const rootNodes = getGgRootNodeIds(program);
    const terminalNodes = getGgTerminalNodeIds(program);

    // Eventuals are opt-in: only check what was explicitly requested
    const eventuals: NamedPredicate<GgTopologyState>[] = [];
    if (this._wantTerminates) {
      const termSet = new Set(terminalNodes);
      eventuals.push({
        name: 'eventually_terminal',
        test: (s) => termSet.has(s.nodeId),
      });
    }
    if (this._wantCollapses) {
      eventuals.push({
        name: 'eventually_beta1_zero',
        test: (s) => s.beta1 === 0,
      });
    }
    eventuals.push(...this.extraEventuals);

    const checkerOpts: CheckerOptions<GgTopologyState> = {
      maxDepth: this._maxDepth,
      invariants: [
        { name: 'beta1_non_negative', test: (s) => s.beta1 >= 0 },
        { name: 'beta1_lt_bound', test: (s) => s.beta1 < this._maxBeta1 },
        ...this.extraInvariants,
      ],
      eventual: eventuals,
    };

    const raw = await checkGgProgram(normalizedSource, {
      model: this.modelOpts,
      checker: checkerOpts,
    });

    // Reachability checks via BFS on the temporal model
    const reachViolations: GGTestResult['violations'] = [];
    if (this._reachableNodes.length > 0 || this._reachableBeta1.length > 0) {
      const model = buildGgTemporalModel(program, this.modelOpts);
      const visited = new Set<string>();
      const reachedNodes = new Set<string>();
      const reachedBeta1 = new Set<number>();
      let frontier = [...model.initialStates];

      while (frontier.length > 0) {
        const nextFrontier: GgTopologyState[] = [];
        for (const state of frontier) {
          const fp = model.fingerprint(state);
          if (visited.has(fp)) continue;
          visited.add(fp);
          reachedNodes.add(state.nodeId);
          reachedBeta1.add(state.beta1);

          for (const action of model.actions) {
            if (action.enabled && !action.enabled(state)) continue;
            for (const succ of action.successors(state)) {
              if (!visited.has(model.fingerprint(succ))) {
                nextFrontier.push(succ);
              }
            }
          }
        }
        frontier = nextFrontier;
        if (visited.size > 10000) break; // safety limit
      }

      for (const nodeId of this._reachableNodes) {
        if (!reachedNodes.has(nodeId)) {
          reachViolations.push({
            kind: 'eventual',
            name: `reachable_${nodeId}`,
            message: `Node '${nodeId}' is not reachable from initial state`,
          });
        }
      }
      for (const beta1 of this._reachableBeta1) {
        if (!reachedBeta1.has(beta1)) {
          reachViolations.push({
            kind: 'eventual',
            name: `reaches_beta1_${beta1}`,
            message: `beta1=${beta1} is never reached`,
          });
        }
      }
    }

    const allViolations = [
      ...raw.violations.map((v) => ({
        kind: v.kind,
        name: v.name,
        message: v.message,
      })),
      ...reachViolations,
    ];

    return {
      ok: raw.ok && reachViolations.length === 0,
      complete: raw.complete,
      stateCount: raw.stateCount,
      topology: { ...raw.topology },
      program: {
        nodeCount: program.nodes.length,
        edgeCount: program.edges.length,
        rootNodes,
        terminalNodes,
      },
      violations: allViolations,
      raw,
    };
  }
}

// ═══════════════════════════════════════════════════════════════════════════════
// Public API — Entry point for test harness
// ═══════════════════════════════════════════════════════════════════════════════

/** Create a GG test builder for a topology source */
export function ggTest(source: string): GGTestBuilder {
  return new GGTestBuilder(source);
}

/**
 * Quick-check a .gg source with default settings.
 * Returns the full test result.
 */
export async function ggQuickCheck(source: string): Promise<GGTestResult> {
  return ggTest(source).safe().run();
}

/**
 * Assert a .gg topology is well-formed.
 * Throws if any verification fails.
 */
export async function ggAssert(
  source: string,
  label?: string
): Promise<GGTestResult> {
  const result = await ggQuickCheck(source);
  if (!result.ok) {
    const violations = result.violations
      .map((v) => `  [${v.kind}] ${v.name}: ${v.message}`)
      .join('\n');
    throw new Error(
      `GG verification failed${label ? ` for ${label}` : ''}:\n${violations}`
    );
  }
  return result;
}
