// @a0n/.ignored_gnosis/src/gnarly-compiler.ts
/**
 * Gnarly compiler.
 *
 * Lowers `.gnarly` multilingual GG-family source into existing Gnosis
 * topology, polyglot manifest, scaffold, and speed-lint surfaces.
 */

import {
  BettyCompiler,
  type GraphAST,
  type Diagnostic,
} from './betty/compiler';
import {
  buildMultiLanguageManifest,
  type MultiLanguageManifest,
  type PolyglotExecutionManifest,
  type NodeExecutionPlan,
} from './polyglot-execution-bridge';
import {
  findBestLanguage,
  generateTopoRaceTopology,
  scaffoldTopology,
  type ExtractedFunction,
  type LanguageFitness,
} from './polyglot-compose';
import { getPolyglotCapabilityMatrix } from './polyglot-registry';
import {
  PolyglotStrategyMemory,
  type StrategyScore,
} from './polyglot-memory';
import {
  parseGnarly,
  type GnarlyDocument,
  type GnarlyImplementationBlock,
} from './gnarly-parser';

export type GnarlyDiagnosticSeverity = 'error' | 'warning' | 'info' | 'hint';
export type GnarlySpeedEvidence = 'proved' | 'predicted' | 'observed';

export interface GnarlyDiagnostic {
  line: number;
  column: number;
  code: string;
  message: string;
  severity: GnarlyDiagnosticSeverity;
}

export interface GnarlySpeedDiagnostic extends GnarlyDiagnostic {
  code:
    | 'GNARLY_FASTER_LANGUAGE_AVAILABLE'
    | 'GNARLY_FASTER_TOPOLOGY_AVAILABLE'
    | 'GNARLY_RACE_CANDIDATE_AVAILABLE';
  severity: 'hint';
  nodeId: string;
  currentLanguage?: string;
  recommendedLanguage: string;
  evidence: GnarlySpeedEvidence;
  score: number;
  rationale: string;
  recommendations: LanguageFitness[];
  observedScore?: StrategyScore;
}

export interface GnarlyGeneratedFile {
  fileName: string;
  language: string;
  nodeId: string;
  source: string;
  embedded: boolean;
}

export interface GnarlyCompileOptions {
  filePath?: string;
  strategyMemory?: PolyglotStrategyMemory;
  maxRecommendations?: number;
}

export interface GnarlyCompileResult {
  document: GnarlyDocument;
  ggSource: string;
  ast: GraphAST | null;
  output: string;
  diagnostics: GnarlyDiagnostic[];
  speedDiagnostics: GnarlySpeedDiagnostic[];
  executionManifest: PolyglotExecutionManifest;
  multiLanguageManifest: MultiLanguageManifest | null;
  generatedFiles: GnarlyGeneratedFile[];
  topoRaceGg?: string;
}

const LANGUAGE_EXTENSIONS: Record<string, string> = {
  bash: 'sh',
  c: 'c',
  cpp: 'cpp',
  c_sharp: 'cs',
  elixir: 'exs',
  go: 'go',
  haskell: 'hs',
  java: 'java',
  javascript: 'mjs',
  kotlin: 'kt',
  lua: 'lua',
  ocaml: 'ml',
  php: 'php',
  python: 'py',
  ruby: 'rb',
  rust: 'rs',
  scala: 'scala',
  swift: 'swift',
  typescript: 'ts',
  zig: 'zig',
};

function languageExtension(language: string): string {
  return LANGUAGE_EXTENSIONS[language] ?? 'txt';
}

function toSnakeCase(value: string): string {
  return value
    .replace(/([a-z])([A-Z])/g, '$1_$2')
    .replace(/[^A-Za-z0-9]+/g, '_')
    .replace(/^_+|_+$/g, '')
    .toLowerCase();
}

function parseStringList(raw: string | undefined): string[] {
  if (!raw) {
    return [];
  }
  const trimmed = raw.trim();
  if (!trimmed.startsWith('[') || !trimmed.endsWith(']')) {
    return trimmed.length > 0 ? [trimmed.replace(/^['"]|['"]$/g, '')] : [];
  }
  return trimmed
    .slice(1, -1)
    .split(',')
    .map((entry) => entry.trim().replace(/^['"]|['"]$/g, ''))
    .filter((entry) => entry.length > 0);
}

function parseBoolean(raw: string | undefined): boolean {
  return raw === 'true' || raw === '1' || raw === 'yes';
}

function diagnosticFromBetty(diagnostic: Diagnostic): GnarlyDiagnostic {
  return {
    line: diagnostic.line,
    column: diagnostic.column,
    code: diagnostic.code ?? 'GNARLY_GG_DIAGNOSTIC',
    message: diagnostic.message,
    severity: diagnostic.severity,
  };
}

function validateLanguage(
  language: string,
  validLanguages: Set<string>,
  diagnostics: GnarlyDiagnostic[],
  context: string,
  line = 1
): void {
  if (!validLanguages.has(language)) {
    diagnostics.push({
      line,
      column: 1,
      code: 'GNARLY_UNKNOWN_LANGUAGE',
      message: `Unknown Gnarly language '${language}' in ${context}. Registered languages come from the Gnosis polyglot registry.`,
      severity: 'warning',
    });
  }
}

function implementationFileName(block: GnarlyImplementationBlock): string {
  return `${toSnakeCase(block.nodeId)}.${languageExtension(block.language)}`;
}

function buildExecutionManifest(
  ast: GraphAST | null,
  defaultLanguage: string,
  defaultFilePath: string,
  entryFunction: string
): PolyglotExecutionManifest {
  const plans: NodeExecutionPlan[] = [];
  if (ast) {
    for (const [nodeId, node] of ast.nodes) {
      const kind = node.labels.includes('PolyglotBridgeCall')
        ? 'call'
        : node.labels.includes('Entry') || nodeId === entryFunction
        ? 'entry'
        : node.labels.includes('Return')
        ? 'return'
        : 'statement';
      const plan: NodeExecutionPlan = {
        node_id: nodeId,
        source_range: { start_byte: 0, end_byte: 0 },
        kind,
      };
      const callee = node.properties.callee ?? node.properties.name;
      if (callee) {
        plan.callee = callee;
      }
      plans.push(plan);
    }
  }

  return {
    language: defaultLanguage,
    file_path: defaultFilePath,
    entry_function: entryFunction,
    node_execution_plans: plans,
  };
}

function extractFunctionForNode(
  ast: GraphAST,
  nodeId: string,
  sourceLanguage: string,
  filePath: string
): ExtractedFunction {
  const node = ast.nodes.get(nodeId);
  const edgeTypes = ast.edges
    .filter(
      (edge) =>
        edge.sourceIds.includes(nodeId) || edge.targetIds.includes(nodeId)
    )
    .map((edge) => edge.type);
  const callees = node?.properties.callee ? [node.properties.callee] : [];

  return {
    filePath,
    language: sourceLanguage,
    name: node?.properties.callee ?? nodeId,
    params: [],
    callees,
    ast: {
      nodes: new Map(node ? [[nodeId, node]] : []),
      edges: ast.edges.filter(
        (edge) =>
          edge.sourceIds.includes(nodeId) || edge.targetIds.includes(nodeId)
      ),
    },
    ggSource: '',
    nodeCount: Math.max(1, node ? 1 : 0),
    edgeTypes,
  };
}

function buildSpeedDiagnostic(
  nodeId: string,
  currentLanguage: string | undefined,
  recommendations: LanguageFitness[],
  observedScore: StrategyScore | null
): GnarlySpeedDiagnostic | null {
  const top = recommendations[0];
  if (!top) {
    return null;
  }

  const observedWins =
    observedScore && observedScore.language !== currentLanguage
      ? observedScore
      : null;
  const recommendedLanguage = observedWins?.language ?? top.language;
  if (currentLanguage && recommendedLanguage === currentLanguage) {
    return null;
  }

  const evidence: GnarlySpeedEvidence = observedWins ? 'observed' : 'predicted';
  const message =
    evidence === 'observed'
      ? `Gnarly has observed ${recommendedLanguage} running faster for ${nodeId}; open a preview before changing files.`
      : `Gnarly predicts ${recommendedLanguage} is the fastest fit for ${nodeId}; open a preview before changing files.`;

  const diagnostic: GnarlySpeedDiagnostic = {
    line: 1,
    column: 1,
    code: 'GNARLY_FASTER_LANGUAGE_AVAILABLE',
    message,
    severity: 'hint',
    nodeId,
    recommendedLanguage,
    evidence,
    score: observedWins?.fitnessScore ?? top.score,
    rationale: observedWins
      ? `${recommendedLanguage}: ${observedWins.observations} observation(s), ${observedWins.weightedDurationMs.toFixed(1)}ms weighted mean.`
      : top.rationale,
    recommendations,
  };

  if (currentLanguage) {
    diagnostic.currentLanguage = currentLanguage;
  }
  if (observedWins) {
    diagnostic.observedScore = observedWins;
  }
  return diagnostic;
}

function buildGeneratedFiles(
  ast: GraphAST | null,
  document: GnarlyDocument,
  speedDiagnostics: GnarlySpeedDiagnostic[]
): GnarlyGeneratedFile[] {
  const generatedFiles: GnarlyGeneratedFile[] = document.implementations.map(
    (block) => ({
      fileName: implementationFileName(block),
      language: block.language,
      nodeId: block.nodeId,
      source: block.sourceText,
      embedded: true,
    })
  );

  if (!ast || speedDiagnostics.length === 0) {
    return generatedFiles;
  }

  const assignments = speedDiagnostics.map((diagnostic) => ({
    nodeId: diagnostic.nodeId,
    language: diagnostic.recommendedLanguage,
    functionName:
      ast.nodes.get(diagnostic.nodeId)?.properties.callee ?? diagnostic.nodeId,
  }));
  const scaffold = scaffoldTopology(ast, assignments, document.ggSource);
  for (const file of scaffold.files) {
    generatedFiles.push({
      fileName: file.fileName,
      language: file.language,
      nodeId: file.nodeId,
      source: file.source,
      embedded: false,
    });
  }

  return generatedFiles;
}

/**
 * Handles the gnosis compile Gnarly workflow.
 */
export async function compileGnarly(
  source: string,
  options: GnarlyCompileOptions = {}
): Promise<GnarlyCompileResult> {
  const filePath = options.filePath ?? 'inline.gnarly';
  const document = parseGnarly(source, filePath);
  const diagnostics: GnarlyDiagnostic[] = document.diagnostics.map(
    (diagnostic) => ({
      ...diagnostic,
      code: 'GNARLY_PARSE_DIAGNOSTIC',
    })
  );

  const compiler = new BettyCompiler();
  const bettyResult = compiler.parse(document.ggSource);
  diagnostics.push(...bettyResult.diagnostics.map(diagnosticFromBetty));

  const matrix = await getPolyglotCapabilityMatrix();
  const validLanguages = new Set(matrix.languages.map((language) => language.id));
  for (const language of document.metadata.languages) {
    validateLanguage(language, validLanguages, diagnostics, 'metadata languages');
  }
  for (const block of document.implementations) {
    validateLanguage(
      block.language,
      validLanguages,
      diagnostics,
      `implementation block for ${block.nodeId}`,
      block.startLine
    );
  }

  const ast = bettyResult.ast;
  const defaultLanguage = document.metadata.languages[0] ?? 'typescript';
  const entryFunction =
    document.metadata.entry ??
    (ast ? [...ast.nodes.keys()][0] : undefined) ??
    'main';
  const defaultFilePath = filePath;
  const speedDiagnostics: GnarlySpeedDiagnostic[] = [];
  let topoRaceGg: string | undefined;

  if (ast) {
    for (const [nodeId, node] of ast.nodes) {
      const currentLanguage = node.properties.language ?? defaultLanguage;
      if (node.properties.language) {
        validateLanguage(
          node.properties.language,
          validLanguages,
          diagnostics,
          `node ${nodeId}`
        );
      }

      const candidateLanguages = parseStringList(node.properties.candidates);
      for (const language of candidateLanguages) {
        validateLanguage(language, validLanguages, diagnostics, `node ${nodeId} candidates`);
      }

      const wantsFastest =
        parseBoolean(node.properties.fastest) || candidateLanguages.length > 0;
      if (!wantsFastest) {
        continue;
      }

      const functionShape = extractFunctionForNode(
        ast,
        nodeId,
        currentLanguage,
        node.properties.file ?? filePath
      );
      const ranked = findBestLanguage(
        functionShape,
        candidateLanguages.length > 0 ? candidateLanguages : undefined
      ).slice(0, options.maxRecommendations ?? 5);
      const observedLanguage = options.strategyMemory?.recommend(nodeId);
      const observedScore = observedLanguage
        ? options.strategyMemory?.score(nodeId, observedLanguage) ?? null
        : null;
      const speedDiagnostic = buildSpeedDiagnostic(
        nodeId,
        currentLanguage,
        ranked,
        observedScore
      );
      if (speedDiagnostic) {
        speedDiagnostics.push(speedDiagnostic);
      }

      if (!topoRaceGg && ranked.length > 0) {
        topoRaceGg = generateTopoRaceTopology(
          node.properties.callee ?? nodeId,
          node.properties.file ?? filePath,
          ranked.slice(0, 3).map((entry) => entry.language)
        );
      }
    }
  }

  const executionManifest = buildExecutionManifest(
    ast,
    defaultLanguage,
    defaultFilePath,
    entryFunction
  );
  const multiLanguageManifest = ast
    ? buildMultiLanguageManifest(ast, defaultLanguage, defaultFilePath)
    : null;

  const result: GnarlyCompileResult = {
    document,
    ggSource: document.ggSource,
    ast,
    output: bettyResult.output,
    diagnostics,
    speedDiagnostics,
    executionManifest,
    multiLanguageManifest,
    generatedFiles: buildGeneratedFiles(ast, document, speedDiagnostics),
  };

  if (topoRaceGg) {
    result.topoRaceGg = topoRaceGg;
  }
  return result;
}

/**
 * Handles the gnosis analyze Gnarly workflow.
 */
export async function analyzeGnarly(
  source: string,
  options: GnarlyCompileOptions = {}
): Promise<GnarlyCompileResult> {
  return compileGnarly(source, options);
}

export {
  parseGnarly,
  type GnarlyDocument,
  type GnarlyImplementationBlock,
} from './gnarly-parser';
export { PolyglotStrategyMemory } from './polyglot-memory';
