// @a0n/.ignored_gnosis/src/gnode-execution.test.ts
import { mkdtempSync, rmSync, writeFileSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { describe, expect, it } from '@a0n/gnosis/test';
import {
  GnosisGnodeExecutionSurface,
  type GnodeExecutionControlResponse,
  type GnodeExecutionEnvelopeStatus,
  type GnodeExecutionResponse,
  type GnodeExecutionSurfaceStatus,
  type SerializedGnodeArtifact,
} from './gnode-execution';
import {
  compileTypeScriptToGnosis,
  type GnosisTypeScriptBridgeResult,
} from './ts-bridge';

function encodeJson(value: unknown): Uint8Array {
  return new TextEncoder().encode(JSON.stringify(value));
}

function decodeJson<T>(response: GnodeExecutionControlResponse): T {
  return JSON.parse(new TextDecoder().decode(response.body)) as T;
}

function serializeCompiled(
  compiled: GnosisTypeScriptBridgeResult
): SerializedGnodeArtifact {
  const { ast: _ast, ...serialized } = compiled;
  return serialized;
}

function compileFixture(sourceText: string): {
  readonly compiled: SerializedGnodeArtifact;
  readonly modulePath: string;
  readonly cleanup: () => void;
} {
  const directory = mkdtempSync(join(tmpdir(), 'gnosis-gnode-'));
  const modulePath = join(directory, 'app.ts');
  writeFileSync(modulePath, sourceText, 'utf8');

  const compiled = compileTypeScriptToGnosis(sourceText, {
    exportName: 'main',
    sourceFilePath: modulePath,
  });

  return {
    compiled: serializeCompiled(compiled),
    modulePath,
    cleanup: () => {
      rmSync(directory, { recursive: true, force: true });
    },
  };
}

describe('GnosisGnodeExecutionSurface', () => {
  it('registers and executes a compiled artifact directly', async () => {
    const fixture = compileFixture(`
      function decorate(userId: string): string {
        return 'user-' + userId;
      }

      export async function main(input: { userId: string }): Promise<string> {
        return decorate(input.userId);
      }
    `);

    try {
      const surface = new GnosisGnodeExecutionSurface({
        laneCount: 3,
        strategy: 'cannon',
      });

      const artifact = surface.registerArtifact({
        compiled: fixture.compiled,
        modulePath: fixture.modulePath,
      });

      const first = await surface.execute({
        artifactId: artifact.artifactId,
        input: { userId: 'ada' },
      });
      const second = await surface.execute({
        artifactId: artifact.artifactId,
        input: { userId: 'ada' },
      });
      const status = surface.getStatus();

      expect(first.payload).toBe('user-ada');
      expect(first.cache?.status).toBe('miss');
      expect(first.plan.strategy).toBe('cannon');
      expect(first.envelope.envelopeId).toHaveLength(64);
      expect(first.envelope.core.appId).toBe('app');
      expect(second.payload).toBe('user-ada');
      expect(second.cache?.status).toBe('hit');
      expect(status.registeredArtifacts).toBe(1);
      expect(status.envelopes[0]?.digest).toBe(first.envelope.digest);
      expect(status.profiles[0]?.executions).toBe(2);
      expect(status.profiles[0]?.lastCacheStatus).toBe('hit');
      expect(status.coreCacheEvents).toBeGreaterThan(0);
    } finally {
      fixture.cleanup();
    }
  });

  it('serves register, execute, and status control routes', async () => {
    const fixture = compileFixture(`
      let active = 0;

      async function guard(label: string): Promise<string> {
        active += 1;
        if (active > 1) {
          active -= 1;
          throw new Error('overlap:' + label);
        }

        await new Promise((resolve) => setTimeout(resolve, 25));
        active -= 1;
        return 'done:' + label;
      }

      export async function main(input: { label: string }): Promise<string> {
        return guard(input.label);
      }
    `);

    try {
      const surface = new GnosisGnodeExecutionSurface({
        enabled: true,
        singleThread: true,
        laneCount: 4,
        strategy: 'cannon',
        runtimeBindingHash:
          'binding-gnode-runtime-0123456789abcdef0123456789abcdef',
      });

      const registerResponse = await surface.handleRequest({
        method: 'POST',
        path: '/.aeon/gnode/register',
        body: encodeJson({
          compiled: fixture.compiled,
          modulePath: fixture.modulePath,
        }),
      });

      expect(registerResponse?.status).toBe(200);
      const registered = decodeJson<{
        artifactId: string;
        envelope: { envelopeId: string; digest: string };
      }>(registerResponse!);
      expect(registered.envelope.envelopeId).toHaveLength(64);

      const [firstResponse, secondResponse] = await Promise.all([
        surface.handleRequest({
          method: 'GET',
          path: `/.aeon/gnode/execute/${registered.artifactId}`,
          query: `?input_json=${encodeURIComponent(
            JSON.stringify({ label: 'alpha' })
          )}`,
          headers: {
            'x-aeon-auth-verified': '1',
            'x-aeon-actor': 'did:key:alice',
            'x-ucan-capabilities': 'gnosis/steering.run,aeon/process',
          },
        }),
        surface.handleRequest({
          method: 'GET',
          path: `/.aeon/gnode/execute/${registered.artifactId}`,
          query: `?input_json=${encodeURIComponent(
            JSON.stringify({ label: 'beta' })
          )}`,
        }),
      ]);

      expect(firstResponse?.status).toBe(200);
      expect(secondResponse?.status).toBe(200);
      const firstExecution = decodeJson<GnodeExecutionResponse>(firstResponse!);
      expect(firstExecution.payload).toBe('done:alpha');
      expect(firstExecution.executionAuth?.enforced).toBe(true);
      expect(firstExecution.executionAuth?.principal).toBe('did:key:alice');
      expect(firstExecution.executionAuth?.capabilityCount).toBe(2);
      expect(firstExecution.executionAuth?.reactorHandleGrantCount).toBe(2);
      expect(firstExecution.executionAuth?.reactorHandleKinds).toEqual([
        'request',
        'response',
      ]);
      expect(firstExecution.executionAuth?.source).toBe('headers');
      expect(decodeJson<GnodeExecutionResponse>(secondResponse!).payload).toBe(
        'done:beta'
      );

      const statusResponse = await surface.handleRequest({
        method: 'GET',
        path: '/.aeon/gnode/status',
      });

      expect(statusResponse?.status).toBe(200);
      const status = decodeJson<GnodeExecutionSurfaceStatus>(statusResponse!);
      expect(status.enabled).toBe(true);
      expect(status.singleThread).toBe(true);
      expect(status.queueDepth).toBe(0);
      expect(status.activeExecutions).toBe(0);
      expect(status.registeredArtifacts).toBe(1);
      expect(
        (status.envelopes as GnodeExecutionEnvelopeStatus[])[0]?.digest
      ).toBe(registered.envelope.digest);
      expect(status.profiles[0]?.executions).toBe(2);
    } finally {
      fixture.cleanup();
    }
  });
});
