// @a0n/.ignored_gnosis/src/gnostic-time.ts
/**
 * gnostic-time.ts -- The Picolorenzo Calendar and Clinamen Ladder
 *
 * Canonical spec (source of truth):
 *   open-source/gnosis/lean/Lean/ForkRaceFoldTheorems/GnosticTime.lean
 *
 * This formalizes the Gnostic Time structure, including the Egyptian/Coptic
 * 12x30+5+1 calendar base, the dual clinamen (leap day vs leap month), and
 * the scale-invariant Clinamen Ladder.
 */

// ============================================================================
// Time units (in millionths of a day for high-precision integer arithmetic)
// 1 day = 1,000,000 microdays.
// ============================================================================

export const PI_MICRODAYS = 3141593n;
export const PICOLORENZO = PI_MICRODAYS;

export const NATURAL_WEEK = 7000000n;
export const NATURAL_LUNAR_MONTH = 29530000n;
export const NATURAL_CALENDAR_MONTH = 30440000n;
export const NATURAL_HALF_YEAR = 182500000n;

// ============================================================================
// Gnostic Time Intervals
// ============================================================================

export const BARBELO_PLO = 1n * PICOLORENZO;
export const SYZYGY_PLO = 2n * PICOLORENZO;
export const PROTON_PLO = 3n * PICOLORENZO;
export const SOPHIA_PLO = 9n * PICOLORENZO;
export const KENOMA_PLO = 10n * PICOLORENZO;
export const VOID_PLO = 21n * PICOLORENZO;
export const PLEROMA_PLO = 55n * PICOLORENZO;

// ============================================================================
// The Clinamen Ladder: Equal Exchange at Every Scale
// ============================================================================

/**
 * The Clinamen tier dictates the RESOLUTION at which the temporal drift
 * The Clinamen tier dictates the RESOLUTION at which the temporal drift
 * is corrected. 
 * - Payment is always EXACTLY 1 (Barbelo) mapped to the selected tier.
 * - Finer resolutions prevent large accumulated debt.
 * - Coarser resolutions defer the cost at the expense of local drift.
 */
export enum ClinamenTier {
  Second = 'second',           // 1 leap second (finest)
  Day = 'day',                 // 1 leap day (medium)
  Month = 'month',             // 1 leap month (coarse)
  Picolorenzo = 'picolorenzo', // 1 leap pLo (π-scale)
}

/**
 * THM-COST-is-BARBELO / THM-RESOLUTION-CHOICE
 * The cost of calendar repair is fixed at Barbelo (1). You choose the tier at
 * which to pay it, but the cost itself never changes.
 */
export function getClinamenCost(_tier: ClinamenTier): bigint {
  return 1n; // Nat.succ 0
}

/**
 * Gets the actual microday representation of the clinamen temporal gap 
 * based on the chosen measurement tier.
 */
export function getClinamenDurationMicrodays(tier: ClinamenTier): bigint {
  switch (tier) {
    case ClinamenTier.Second:
      return 1000000n / 86400n; // 1 second in microdays (approx)
    case ClinamenTier.Day:
      return 1000000n; // 1 day
    case ClinamenTier.Month:
      return NATURAL_LUNAR_MONTH; // 1 lunar month
    case ClinamenTier.Picolorenzo:
      return PICOLORENZO; // 1 pLo
  }
}

// ============================================================================
// Egyptian Calendar Primitives
// ============================================================================

export const EGYPTIAN_IDEAL_DAYS = 360n; // 12 * 30
export const EGYPTIAN_EPAGOMENAL_DAYS = 5n; // fib(6) - fib(4)
export const COPTIC_SOLAR_YEAR = EGYPTIAN_IDEAL_DAYS + EGYPTIAN_EPAGOMENAL_DAYS; // 365
export const COPTIC_LEAP_YEAR = COPTIC_SOLAR_YEAR + getClinamenCost(ClinamenTier.Day); // 366

// ============================================================================
// Circadian Gnosis Alignment
// ============================================================================

/**
 * THM-CIRCADIAN-GNOSIS-ALIGNMENT:
 * The average human circadian rhythm (24.2h) deviates from the solar day (24.0h)
 * by exactly 12 minutes (0.2h). 
 * 
 * Drift = 12m = 1 Aeon minute.
 * Ratio = 12 / 1440 = 1/120 = 1 / (Aeon * Kenoma).
 */
export const SOLAR_DAY_MINUTES = 1440n;
export const CIRCADIAN_DRIFT_MINUTES = 12n; // Aeon
export const HUMAN_CIRCADIAN_MINUTES = SOLAR_DAY_MINUTES + CIRCADIAN_DRIFT_MINUTES; // 1452

/**
 * The fundamental drift ratio of biological life.
 * 1 / (AEON * KENOMA) = 1/120.
 */
export const AEON_KENOMA_GAP_RATIO_DENOMINATOR = 120n;

// ============================================================================
// Resolution Debt Analysis
// ============================================================================

/**
 * THM-EGYPTIAN-DEBT: If you choose resolution = never (no clinamen),
 * the debt accumulates infinitely. 
 * This returns the max deferred debt (in days) allowed before drift exceeds a cycle.
 */
export function calculateSothicDebtLimit(): bigint {
  return EGYPTIAN_IDEAL_DAYS * 4n; // Approx 1440-1460 years. 1461 years to outlap 365 days.
}
