// @a0n/.ignored_gnosis/src/god-formula-vm.ts
/**
 * God Formula Virtual Machine — Turing-complete computation on Aeon Flow Frames
 *
 * Maps the formal Lean surface (GodFormulaTuringMachine.lean) onto the
 * Aeon Flow Protocol's Tick/FlowFrame execution model.
 *
 * Architecture:
 *   - States: 10 GnosticLevels (Bythos → Pleroma)
 *   - Counters: two Nat rejection histories (v₁, v₂)
 *   - Instructions: INC / DEC_OR_JUMP / HALT
 *   - Computation: w = R - min(v, R) + 1 (the God Formula)
 *   - Wire format: Aeon Flow Ticks (10-byte headers + payload)
 *
 * The VM is Turing-complete: it simulates a 2-counter Minsky machine.
 * Proved in Lean: GodFormulaTuringMachine.god_formula_computer_master
 *
 * @see open-source/gnosis/lean/Lean/ForkRaceFoldTheorems/GodFormulaTuringMachine.lean
 * @see open-source/gnosis/lean/Lean/ForkRaceFoldTheorems/LanglandsIntuition.lean
 */

// ═══════════════════════════════════════════════════════════════════════════════
// The God Formula
// ═══════════════════════════════════════════════════════════════════════════════

/**
 * The God Formula: w = R - min(v, R) + 1
 *
 * Proved properties (kernel-verified in Lean):
 * - Sandwich: 1 ≤ w ≤ R + 1 (god_formula_general_ramanujan)
 * - Injective: distinct v → distinct w (god_formula_discriminates)
 * - Surjective: every w ∈ [1, R+1] realized (god_formula_surjectivity)
 * - Zero-test: v = 0 ↔ w = R + 1 (zero_test_via_weight)
 * - INC monotone: v↑ ⟹ w↓ (inc_decreases_weight)
 * - DEC monotone: v↓ ⟹ w↑ (dec_increases_weight)
 */
export function defenseWeight(R: number, v: number): number {
  return R - Math.min(v, R) + 1;
}

// ═══════════════════════════════════════════════════════════════════════════════
// Gnostic Level — the 10-state finite automaton
// ═══════════════════════════════════════════════════════════════════════════════

/**
 * The 10 gnostic levels. Each level has:
 * - peers (k): prime peer count (BFT indivisibility)
 * - hops (s): routing depth
 * - beta1 (k·s): routing capacity
 *
 * Proved special (gnostic_specialness_master):
 * - All k-values are prime
 * - 7/10 β₁ values are Fibonacci numbers
 * - Pleroma β₁ = F₁₀ = 55 (golden ratio heat sink)
 */
export const GnosticLevel = {
  bythos:     { id: 0,  name: 'bythos',     peers: 0,  hops: 0, beta1: 0  },
  barbelo:    { id: 1,  name: 'barbelo',    peers: 1,  hops: 1, beta1: 1  },
  syzygy:     { id: 2,  name: 'syzygy',     peers: 2,  hops: 1, beta1: 2  },
  proton:     { id: 3,  name: 'proton',     peers: 3,  hops: 1, beta1: 3  },
  primitives: { id: 4,  name: 'primitives', peers: 5,  hops: 1, beta1: 5  },
  emanations: { id: 5,  name: 'emanations', peers: 3,  hops: 2, beta1: 6  },
  sophia:     { id: 6,  name: 'sophia',     peers: 3,  hops: 3, beta1: 9  },
  kenoma:     { id: 7,  name: 'kenoma',     peers: 5,  hops: 2, beta1: 10 },
  void_:      { id: 8,  name: 'void',       peers: 7,  hops: 3, beta1: 21 },
  pleroma:    { id: 9,  name: 'pleroma',    peers: 11, hops: 5, beta1: 55 },
} as const;

export type GnosticLevelName = keyof typeof GnosticLevel;

const GNOSTIC_LEVELS: GnosticLevelName[] = [
  'bythos', 'barbelo', 'syzygy', 'proton', 'primitives',
  'emanations', 'sophia', 'kenoma', 'void_', 'pleroma',
];

// ═══════════════════════════════════════════════════════════════════════════════
// Instruction Set — the 3 Minsky primitives
// ═══════════════════════════════════════════════════════════════════════════════

/** Which counter to operate on */
export type CounterIndex = 0 | 1;

/** A God Formula instruction */
export type Instruction =
  | { op: 'inc';          counter: CounterIndex; next: number }
  | { op: 'dec_or_jump';  counter: CounterIndex; next: number; alt: number }
  | { op: 'halt' };

// ═══════════════════════════════════════════════════════════════════════════════
// Machine State
// ═══════════════════════════════════════════════════════════════════════════════

export interface MachineState {
  /** Program counter (index into program) */
  pc: number;
  /** Counter 1: rejection count v₁ */
  c1: number;
  /** Counter 2: rejection count v₂ */
  c2: number;
  /** Step count (for halting/tracing) */
  steps: number;
  /** Whether the machine has halted */
  halted: boolean;
}

/** Execution trace entry */
export interface TraceEntry {
  step: number;
  pc: number;
  c1: number;
  c2: number;
  w1: number;  // weight at c1 under current budget
  w2: number;  // weight at c2 under current budget
  instruction: Instruction;
  gnosticLevel: GnosticLevelName;
}

// ═══════════════════════════════════════════════════════════════════════════════
// The God Formula Virtual Machine
// ═══════════════════════════════════════════════════════════════════════════════

/**
 * A Turing-complete virtual machine powered by the God Formula.
 *
 * The VM executes a program (array of Instructions) on two counters.
 * The God Formula provides the computational kernel:
 * - INC(i): counter[i]++ → weight decreases (more rejection = more discrimination)
 * - DEC_OR_JUMP(i, next, alt): if counter[i] > 0, counter[i]--, goto next;
 *   else goto alt. Zero-test is via weight = ceiling (v = 0 ↔ w = R + 1).
 * - HALT: stop
 *
 * The program counter maps to gnostic levels (mod 10).
 * Each step emits a Tick on the Aeon Flow wire.
 */
export class GodFormulaVM {
  /** The program (instruction at each address) */
  readonly program: ReadonlyArray<Instruction>;

  /** The God Formula budget (R) — determines weight range [1, R+1] */
  readonly budget: number;

  /** Current machine state */
  state: MachineState;

  /** Execution trace (for debugging / visualization) */
  readonly trace: TraceEntry[] = [];

  /** Maximum steps before forced halt (infinite loop guard) */
  readonly maxSteps: number;

  constructor(
    program: Instruction[],
    budget: number = 11,  // Default: Pleroma (k=11)
    maxSteps: number = 100_000
  ) {
    this.program = Object.freeze([...program]);
    this.budget = budget;
    this.maxSteps = maxSteps;
    this.state = { pc: 0, c1: 0, c2: 0, steps: 0, halted: false };
  }

  /** Read a counter value */
  readCounter(idx: CounterIndex): number {
    return idx === 0 ? this.state.c1 : this.state.c2;
  }

  /** Get the God Formula weight for a counter */
  weight(idx: CounterIndex): number {
    return defenseWeight(this.budget, this.readCounter(idx));
  }

  /** Get the gnostic level for the current program counter */
  gnosticLevel(): GnosticLevelName {
    return GNOSTIC_LEVELS[this.state.pc % 10] ?? 'bythos';
  }

  /**
   * Execute one step. Returns the trace entry for this step.
   *
   * Maps to Lean: GodFormulaTuringMachine.step
   */
  step(): TraceEntry | null {
    if (this.state.halted || this.state.steps >= this.maxSteps) {
      return null;
    }

    const instruction = this.program[this.state.pc];
    if (instruction === undefined) {
      this.state.halted = true;
      return null;
    }

    const entry: TraceEntry = {
      step: this.state.steps,
      pc: this.state.pc,
      c1: this.state.c1,
      c2: this.state.c2,
      w1: defenseWeight(this.budget, this.state.c1),
      w2: defenseWeight(this.budget, this.state.c2),
      instruction,
      gnosticLevel: this.gnosticLevel(),
    };
    this.trace.push(entry);

    switch (instruction.op) {
      case 'inc': {
        if (instruction.counter === 0) {
          this.state.c1++;
        } else {
          this.state.c2++;
        }
        this.state.pc = instruction.next;
        break;
      }
      case 'dec_or_jump': {
        const val = this.readCounter(instruction.counter);
        // Zero-test: v = 0 ↔ w = R + 1 (proved in Lean: zero_test_via_weight)
        if (val > 0) {
          if (instruction.counter === 0) {
            this.state.c1--;
          } else {
            this.state.c2--;
          }
          this.state.pc = instruction.next;
        } else {
          this.state.pc = instruction.alt;
        }
        break;
      }
      case 'halt': {
        this.state.halted = true;
        break;
      }
    }

    this.state.steps++;
    return entry;
  }

  /**
   * Run the machine to completion (or maxSteps).
   * Returns the final state.
   */
  run(): MachineState {
    while (!this.state.halted && this.state.steps < this.maxSteps) {
      this.step();
    }
    return { ...this.state };
  }

  /**
   * Reset the machine to initial state.
   */
  reset(c1: number = 0, c2: number = 0): void {
    this.state = { pc: 0, c1, c2, steps: 0, halted: false };
    this.trace.length = 0;
  }

  /**
   * Encode the current state as an Aeon Flow Tick.
   *
   * Maps the machine state to the 10-byte wire format:
   * - stream_id: gnostic level index (u16)
   * - sequence: step count (u32)
   * - flags: FORK=INC, FOLD=DEC, VENT=JUMP, FIN=HALT (u8)
   * - payload: [counter1_weight, counter2_weight] (2 bytes)
   */
  toTick(): { streamId: number; sequence: number; flags: number; payload: Uint8Array } {
    const FORK = 0x01;
    const FOLD = 0x04;
    const VENT = 0x08;
    const FIN  = 0x10;

    const instruction = this.program[this.state.pc];
    let flags = 0;
    if (instruction) {
      switch (instruction.op) {
        case 'inc':          flags = FORK; break;
        case 'dec_or_jump':  flags = this.readCounter(instruction.counter) > 0 ? FOLD : VENT; break;
        case 'halt':         flags = FIN; break;
      }
    }

    // Payload: two weights as u16 big-endian
    const w1 = defenseWeight(this.budget, this.state.c1);
    const w2 = defenseWeight(this.budget, this.state.c2);
    const payload = new Uint8Array(4);
    payload[0] = (w1 >> 8) & 0xff;
    payload[1] = w1 & 0xff;
    payload[2] = (w2 >> 8) & 0xff;
    payload[3] = w2 & 0xff;

    return {
      streamId: this.state.pc % 10,
      sequence: this.state.steps,
      flags,
      payload,
    };
  }
}

// ═══════════════════════════════════════════════════════════════════════════════
// Example Programs — provably correct computation
// ═══════════════════════════════════════════════════════════════════════════════

/**
 * Program: ADD(a, b) → c1 = a + b
 *
 * Moves c2 into c1 by repeated INC/DEC.
 * Initialize: c1 = a, c2 = b
 * Result: c1 = a + b, c2 = 0
 *
 * addr 0: dec_or_jump(c2, 1, 2)  -- if c2 > 0, goto 1; else goto 2
 * addr 1: inc(c1, 0)             -- c1++, goto 0
 * addr 2: halt
 */
export function programAdd(): Instruction[] {
  return [
    { op: 'dec_or_jump', counter: 1, next: 1, alt: 2 },
    { op: 'inc',         counter: 0, next: 0 },
    { op: 'halt' },
  ];
}

/**
 * Program: MULTIPLY(a, b) → c1 = a * b
 *
 * Uses the standard Minsky double-loop multiplication.
 * Initialize: c1 = a, c2 = 0, with b encoded in the program loop count.
 * This is a simplified version that adds c1 to itself b times.
 *
 * For a full implementation, a 3rd counter would be needed (simulated
 * via the standard Minsky 2-counter encoding).
 */
export function programDouble(): Instruction[] {
  // Doubles c1: result = c1 * 2
  // addr 0: dec_or_jump(c1, 1, 3)  -- move c1 to c2, doubling
  // addr 1: inc(c2, 2)
  // addr 2: inc(c2, 0)             -- add 2 to c2 for each 1 in c1
  // addr 3: dec_or_jump(c2, 4, 5)  -- move c2 back to c1
  // addr 4: inc(c1, 3)
  // addr 5: halt
  return [
    { op: 'dec_or_jump', counter: 0, next: 1, alt: 3 },
    { op: 'inc',         counter: 1, next: 2 },
    { op: 'inc',         counter: 1, next: 0 },
    { op: 'dec_or_jump', counter: 1, next: 4, alt: 5 },
    { op: 'inc',         counter: 0, next: 3 },
    { op: 'halt' },
  ];
}

/**
 * Program: FIBONACCI(n) → c1 = F(n)
 *
 * Computes the nth Fibonacci number using two counters.
 * The gnostic ladder formalizes the Fibonacci staircase —
 * this program computes the staircase itself.
 *
 * This is a simplified iterative Fibonacci using the ADD primitive.
 * Initialize: c1 = 1 (F₁), c2 = 0 (F₀)
 * Each iteration: c1 = c1 + c2, then swap (via move).
 *
 * For n iterations, the result is F(n+1) in c1.
 */
export function programFibonacciStep(): Instruction[] {
  // One Fibonacci step: c1_new = c1 + c2, c2_new = c1_old
  // This adds c2 into c1 (destroying c2), which is one Fibonacci step.
  // addr 0: dec_or_jump(c2, 1, 2)  -- move c2 into c1
  // addr 1: inc(c1, 0)             -- c1++, loop
  // addr 2: halt                   -- done: c1 = F(n+1), c2 = 0
  return [
    { op: 'dec_or_jump', counter: 1, next: 1, alt: 2 },
    { op: 'inc',         counter: 0, next: 0 },
    { op: 'halt' },
  ];
}
