// @a0n/.ignored_gnosis/src/gravitational-dilation.ts
/**
 * Gravitational Time Dilation — Schwarzschild metric near massive bodies.
 *
 * In general relativity, clocks closer to a massive body tick slower
 * relative to clocks farther away. The Schwarzschild metric gives the
 * exact dilation factor for a non-rotating, uncharged spherical mass:
 *
 *   dτ/dt = √(1 - r_s / r)
 *
 * where r_s = 2GM/c² is the Schwarzschild radius and r is the distance
 * from the center of mass. At the event horizon (r = r_s), time stops
 * entirely for the distant observer. At infinity, dτ/dt → 1.
 *
 * This module computes dilation factors, generates concentric ring
 * fields around gravitational bodies, and combines weak-field
 * contributions from multiple bodies.
 *
 * @packageDocumentation
 */

import { SCHWARZSCHILD_RADIUS_SOLAR_LY } from './math';

// ─── Types ─────────────────────────────────────────────────────────

export interface GravitationalBody {
  readonly id: string;
  readonly label: string;
  readonly massSolar: number;
  readonly positionLy: { readonly x: number; readonly y: number; readonly z: number };
}

export interface DilationRing {
  readonly radiusLy: number;
  readonly radiusSchwarzschildRadii: number;
  readonly dilationFactor: number;
  readonly timePassesAt: string;
}

export interface GravitationalDilationField {
  readonly source: 'GnosisGravitationalDilation';
  readonly body: GravitationalBody;
  readonly schwarzschildRadiusLy: number;
  readonly eventHorizonRadiusLy: number;
  readonly rings: readonly DilationRing[];
}

export interface DilationAtPoint {
  readonly bodyId: string;
  readonly distanceLy: number;
  readonly dilationFactor: number;
  readonly timeSlowdownPercent: number;
}

// ─── Functions ─────────────────────────────────────────────────────

/**
 * Schwarzschild radius in light-years for a body of given solar masses.
 * r_s = massSolar * SCHWARZSCHILD_RADIUS_SOLAR_LY
 */
export function schwarzschildRadius(massSolar: number): number {
  return massSolar * SCHWARZSCHILD_RADIUS_SOLAR_LY;
}

/**
 * Gravitational dilation factor dτ/dt = √(1 - r_s / r).
 * Returns 0 if distance <= r_s (at or inside the event horizon).
 * Approaches 1 as distance → ∞.
 */
export function gravitationalDilationFactor(massSolar: number, distanceLy: number): number {
  const rs = schwarzschildRadius(massSolar);
  if (distanceLy <= rs) {
    return 0;
  }
  return Math.sqrt(1 - rs / distanceLy);
}

/**
 * Format a dilation factor as a human-readable speed string.
 */
function formatDilation(factor: number): string {
  if (factor <= 0) {
    return 'time stopped';
  }
  const percent = factor * 100;
  if (percent >= 99.95) {
    return `${percent.toFixed(4)}% speed`;
  }
  if (percent >= 99.5) {
    return `${percent.toFixed(2)}% speed`;
  }
  if (percent >= 9.5) {
    return `${percent.toFixed(1)}% speed`;
  }
  return `${percent.toFixed(2)}% speed`;
}

/**
 * Generate a dilation field of concentric rings around a gravitational body.
 * Rings are logarithmically spaced from 1.01·r_s out to 1000·r_s.
 */
export function dilationField(body: GravitationalBody, ringCount: number = 10): GravitationalDilationField {
  const rs = schwarzschildRadius(body.massSolar);

  const minMultiplier = 1.01;
  const maxMultiplier = 1000;
  const logMin = Math.log(minMultiplier);
  const logMax = Math.log(maxMultiplier);

  const rings: DilationRing[] = [];
  for (let i = 0; i < ringCount; i++) {
    const t = ringCount === 1 ? 0 : i / (ringCount - 1);
    const multiplier = Math.exp(logMin + t * (logMax - logMin));
    const radiusLy = multiplier * rs;
    const factor = gravitationalDilationFactor(body.massSolar, radiusLy);

    rings.push({
      radiusLy,
      radiusSchwarzschildRadii: multiplier,
      dilationFactor: factor,
      timePassesAt: formatDilation(factor),
    });
  }

  return {
    source: 'GnosisGravitationalDilation',
    body,
    schwarzschildRadiusLy: rs,
    eventHorizonRadiusLy: rs,
    rings,
  };
}

/**
 * Single-point dilation query for a body of given solar masses.
 */
export function dilationAtDistance(massSolar: number, distanceLy: number): DilationAtPoint {
  const factor = gravitationalDilationFactor(massSolar, distanceLy);
  return {
    bodyId: '',
    distanceLy,
    dilationFactor: factor,
    timeSlowdownPercent: (1 - factor) * 100,
  };
}

/**
 * Combined dilation in the weak-field approximation.
 * The product of individual dilation factors from each body at the given point.
 */
export function combinedDilation(
  bodies: readonly GravitationalBody[],
  pointLy: { readonly x: number; readonly y: number; readonly z: number },
): number {
  let product = 1;
  for (const body of bodies) {
    const dx = pointLy.x - body.positionLy.x;
    const dy = pointLy.y - body.positionLy.y;
    const dz = pointLy.z - body.positionLy.z;
    const distance = Math.sqrt(dx * dx + dy * dy + dz * dz);
    product *= gravitationalDilationFactor(body.massSolar, distance);
  }
  return product;
}
