// @a0n/.ignored_gnosis/src/hft-advisory-guard.ts
import {
  createSyntheticVisibleObservation,
  deriveAdvisoryAllocationKernel,
  type AdvisoryAllocationKernelBundle,
  type AdvisorySafeStateReason,
} from './advisory-allocation-kernel';

export const HFT_ADVISORY_PRICE_SCALE = 10_000;
export const HFT_ADVISORY_BUY_CANDIDATE_ID = 1;
export const HFT_ADVISORY_SELL_CANDIDATE_ID = 2;

export type HftAdvisorySide = 'buy' | 'sell';

export interface HftAdvisoryGuardInput {
  readonly source?: string;
  readonly side: HftAdvisorySide;
  readonly confidence: number;
  readonly obstructionIntensity: number;
  readonly bid: number;
  readonly ask?: number;
  readonly lastBid?: number;
  readonly tickTimestampMs: number;
  readonly nowMs?: number;
  readonly loopElapsedMs?: number;
  readonly bufferLength: number;
  readonly minimumBufferLength?: number;
  readonly displayedDepth?: number;
  readonly inventoryExposure: number;
  readonly inventoryLimit: number;
  readonly openOrderCount?: number;
  readonly openOrderLimit?: number;
  readonly deadlineMs?: number;
  readonly maxObservationAgeMs?: number;
  readonly decayPerMs?: number;
  readonly commitNumerator?: number;
  readonly commitDenominator?: number;
  readonly modelSolved?: boolean;
  readonly ingestLocked?: boolean;
  readonly confidenceFloor?: number;
  readonly priceMoveThreshold?: number;
}

export interface HftAdvisoryGuardAggregate {
  readonly observationAgeMs: number;
  readonly loopElapsedMs: number;
  readonly bufferReady: boolean;
  readonly confidenceScore: number;
  readonly confidenceFloor: number;
  readonly lastBid: number;
  readonly priceMoveThreshold: number;
  readonly displayedDepth: number;
}

export interface HftAdvisoryGuardBundle {
  readonly execute: boolean;
  readonly side: HftAdvisorySide;
  readonly candidateId: number;
  readonly blockReason: AdvisorySafeStateReason | null;
  readonly advisory: AdvisoryAllocationKernelBundle;
  readonly aggregate: HftAdvisoryGuardAggregate;
}

function clampNonNegativeInt(value: number): number {
  if (!Number.isFinite(value)) {
    return 0;
  }

  return Math.max(0, Math.round(value));
}

function clampUnitInterval(value: number): number {
  if (!Number.isFinite(value)) {
    return 0;
  }

  return Math.min(1, Math.max(0, value));
}

function toScaledInt(value: number, scale = HFT_ADVISORY_PRICE_SCALE): number {
  if (!Number.isFinite(value)) {
    return 0;
  }

  return Math.max(0, Math.round(value * scale));
}

/**
 * Handles the gnosis hft Candidate Id For Side workflow.
 */
export function hftCandidateIdForSide(side: HftAdvisorySide): number {
  return side === 'buy'
    ? HFT_ADVISORY_BUY_CANDIDATE_ID
    : HFT_ADVISORY_SELL_CANDIDATE_ID;
}

/**
 * Handles the gnosis hft Side For Candidate Id workflow.
 */
export function hftSideForCandidateId(candidateId: number): HftAdvisorySide | null {
  if (candidateId === HFT_ADVISORY_BUY_CANDIDATE_ID) {
    return 'buy';
  }

  if (candidateId === HFT_ADVISORY_SELL_CANDIDATE_ID) {
    return 'sell';
  }

  return null;
}

/**
 * Handles the gnosis derive Hft Advisory Guard workflow.
 */
export function deriveHftAdvisoryGuard(
  input: HftAdvisoryGuardInput
): HftAdvisoryGuardBundle {
  const side = input.side;
  const nowMs = clampNonNegativeInt(input.nowMs ?? Date.now());
  const tickTimestampMs = clampNonNegativeInt(input.tickTimestampMs);
  const observationAgeMs = Math.max(0, nowMs - tickTimestampMs);
  const loopElapsedMs = clampNonNegativeInt(input.loopElapsedMs ?? 0);
  const minimumBufferLength = clampNonNegativeInt(input.minimumBufferLength ?? 15);
  const bufferReady = clampNonNegativeInt(input.bufferLength) >= minimumBufferLength;
  const confidenceScore = Math.round(clampUnitInterval(input.confidence) * 1000);
  const confidenceFloor = clampNonNegativeInt(
    input.confidenceFloor ??
      Math.min(995, 650 + Math.round(clampUnitInterval(input.obstructionIntensity) * 200))
  );
  const lastBidInput = input.lastBid;
  const lastBid =
    typeof lastBidInput === 'number' && Number.isFinite(lastBidInput)
      ? lastBidInput
      : input.bid;
  const spread = Math.abs((input.ask ?? input.bid) - input.bid);
  const velocityMove = Math.abs(input.bid - lastBid);
  const priceMoveThreshold =
    Number.isFinite(input.priceMoveThreshold) && (input.priceMoveThreshold ?? 0) >= 0
      ? (input.priceMoveThreshold as number)
      : Math.max(spread, velocityMove);
  const candidateId = hftCandidateIdForSide(side);
  const advisory = deriveAdvisoryAllocationKernel({
    source: input.source?.trim() || 'hft-advisory-guard',
    plan: {
      pointIds: ['market-data', 'decision-lane'],
      samplesPerPoint: 1,
      maxAttemptsPerPoint: 1,
    },
    points: [
      {
        pointId: 'market-data',
        accepted: bufferReady ? 1 : 0,
        rejected: bufferReady ? 0 : 1,
        observed: 1,
      },
      {
        pointId: 'decision-lane',
        accepted: confidenceScore > 0 ? 1 : 0,
        rejected: confidenceScore > 0 ? 0 : 1,
        observed: 1,
      },
    ],
    modelSolved: input.modelSolved ?? confidenceScore > 0,
    ingestLocked: input.ingestLocked ?? true,
    frame: {
      observation: createSyntheticVisibleObservation(confidenceScore, confidenceFloor),
      currentPrice: toScaledInt(input.bid),
      lastPrice: toScaledInt(lastBid),
      displayedDepth: clampNonNegativeInt(input.displayedDepth ?? 1),
      elapsedSinceUpdateMs: observationAgeMs,
      maxObservationAgeMs: clampNonNegativeInt(input.maxObservationAgeMs ?? 250),
      decayPerMs: clampNonNegativeInt(input.decayPerMs ?? 1),
      priceMoveThreshold: toScaledInt(priceMoveThreshold),
    },
    candidateIds: [candidateId],
    risk: {
      inventoryExposure: toScaledInt(input.inventoryExposure, 100),
      inventoryLimit: toScaledInt(input.inventoryLimit, 100),
      openOrderCount: clampNonNegativeInt(input.openOrderCount ?? 0),
      openOrderLimit: clampNonNegativeInt(input.openOrderLimit ?? 1),
    },
    budget: {
      deadlineMs: clampNonNegativeInt(input.deadlineMs ?? 25),
      elapsedMs: loopElapsedMs,
      commitNumerator: clampNonNegativeInt(input.commitNumerator ?? 9),
      commitDenominator: clampNonNegativeInt(input.commitDenominator ?? 10),
    },
  });

  return {
    execute: advisory.decision.kind === 'allocate',
    side,
    candidateId,
    blockReason:
      advisory.decision.kind === 'safe-state' ? advisory.decision.reason : null,
    advisory,
    aggregate: {
      observationAgeMs,
      loopElapsedMs,
      bufferReady,
      confidenceScore,
      confidenceFloor,
      lastBid,
      priceMoveThreshold,
      displayedDepth: clampNonNegativeInt(input.displayedDepth ?? 1),
    },
  };
}
