// @a0n/.ignored_gnosis/src/index.ts
import * as fs from 'fs';
import * as path from 'path';
import { execFile, spawnSync } from 'node:child_process';
import { promisify } from 'node:util';
import { fileURLToPath } from 'url';
import { Pipeline } from '@a0n/aeon-pipelines';
import * as twokeys from '@a0n/twokeys';
import {
  BettyCompiler,
  type Diagnostic,
  type GraphAST,
} from './betty/compiler';
export {
  compileLegaleseToGg,
  lowerLegaleseToGg,
  segmentLegalese,
} from './betty/legalese';
import { generateLeanFromGnosisAst } from './betty/lean';
import { GnosisRegistry } from './runtime/registry';
import { GnosisEngine } from './runtime/engine';
import { GnosisCoreCache } from './runtime/core-cache';
import { GnosisNativeRuntime } from './runtime/native-runtime';
import { ModManager } from './mod/manager';
export * from './dimensional-scheduling';
export * from './spiderweb-topology';
export {
  ARM_A,
  ARM_B,
  ARM_C,
  CLOCK_PERIOD,
  EMANATION_SHELLS,
  LIGHT_ORDINAL,
  OLD_TRIANGLE_BARS,
  PHYLE_BARS,
  PHYLE_COMPOSITIONS as PHYLE_CELL_COMPOSITIONS,
  PHYLE_DIRECTIONS,
  PHYLE_MARGIN,
  PHYLE_RAYS,
  PhyleCell,
  PhyleMesh,
  TRIPOD_ARMS,
  createPhyleCell,
  createPhyleMesh,
  resetVertexIds,
  type PhyleBond,
  type PhylePosition,
  type PhyleVertex,
} from './phyle-cell';
export * from './phyle-temporal-depth';
export * from './phyle-comic-parallax';
export * from './dark-fork-topology';
export * from './dark-atomized-encryption';
export * from './dark-sector-safe';
export * from './sentience';

export * from './triangulation-dimensional-coercion';
export * from './wallington-rotation';
export * from './aperiodic-scheduling';
export * from './randomness';
export * from './probability-certificate';
export * from './probability-routing';
export * from './weather-mesh-adapter';
export * from './StressEnergyMonitor';
export * from './CrossPollinationEngine';
export * from './gnostic-time';
export * from './lucas-bridge';
export * from './semiotic-router';
export { QuantumRuntime, type EdgeFabricConfig, type QuantumRuntimeOp, type QuantumRuntimeResult } from './delayed-projection-runtime';
export {
  type EmotionalState,
  type PhysicalEmotionMetrics,
  computePhysicalEmotion,
  getZenState,
} from './physical-emotion';
export {
  type GnosisEmotion,
  type AffectValence,
  type AffectArousal,
  type EmotionTier,
  GnosisEmotionTaxonomy,
} from './emotion-taxonomy';
import {
  detectModuleFormat,
  loadGnosisModuleFromFile,
  parseMGG,
  renderMGG,
} from './mod/loader';
import {
  analyzeGnosisSource,
  finishSteeringTelemetry,
  DEFAULT_GNOSIS_STEERING_MODE,
  formatGnosisViolations,
  startSteeringTelemetry,
  surfaceSteeringMetrics,
  VALID_GNOSIS_STEERING_MODES,
  withSteeringTelemetry,
} from './analysis';
import {
  authorizeBoundaryWalkRun,
  createSteeringTraceRecord,
  DEFAULT_GNOSIS_STEERING_TRACE_WINDOW,
  defaultSteeringTraceRoomName,
  fingerprintSteeringTopology,
  formatSteeringTraceAlert,
  formatSteeringTraceStoreStatus,
  formatSteeringTraceSummary,
  GnosisSteeringTraceStore,
} from './steering-trace';
import type { GnosisSteeringMetrics, GnosisSteeringMode } from './analysis';
import type { RuntimeTarget } from './capabilities/index';
import {
  analyzeTypeScriptTargets,
  parseTsSonarThresholds,
} from './ts-sonar';
import { ggReportToSarif, tsReportToSarif } from './sarif';
import {
  buildGnosisGgMutationReport,
  buildGnosisTypeScriptMutationReport,
} from './testing/mutation-report';
import { generateTlaFromGnosisSource } from './tla-bridge';

import { GnosisNeo4jBridge } from './neo4j-bridge';
import { GnosisFormatter } from './formatter';
import { registerBettiHandlers } from './betti/handlers';
import { renderWithTopologyCompat } from './runtime/renderer-compat';
import {
  authorizeSteeringApply,
  bootstrapExecutionAuthFromTopology,
  normalizeExecutionAuthContext,
  registerCoreAuthHandlers,
} from './auth/index';
import type { GnosisExecutionAuthContext } from './auth/index';
import {
  NeuralEngine,
  GPUEngine,
  WebGPUEngine,
  WebNNEngine,
  Translator,
  NeuronRepository,
  SynapseRepository,
  TOPIC_DOMAIN_TRANSFORMER_TOPOLOGY_FILE,
  TOPIC_DOMAIN_TRANSFORMER_TOPOLOGY,
  TRANSFORMER_HELLO_WORLD_TOPOLOGY,
  getTopicDomainTransformerTopologySource,
  init as initNeuralEngine,
} from './neural-compat';
import { initGnosisTracingFromEnv } from './apm';
import type {
  Neuron,
  Synapse,
  AdapterTrainingConfig,
  NeuralGraphData,
  LoadTopologyOptions,
  NeuralHeteroFabricOptions,
} from './neural-compat';

type TwokeysRuntime = {
  Series: new (options?: { data?: number[] }) => {
    mean: () => number;
    median: () => { datum: number };
    outliers: () => unknown;
    describe: () => unknown;
  };
  Points: new (options?: { data?: number[][] }) => {
    centroid: () => unknown;
    describe: () => unknown;
  };
  pageRank: (nodes: unknown[], edges: unknown[]) => unknown;
  louvainCommunities: (nodes: unknown[], edges: unknown[]) => unknown;
  graphEda: (nodes: unknown[], edges: unknown[]) => unknown;
};

const twokeysRuntime = twokeys as unknown as TwokeysRuntime;

export {
  GnosisNeo4jBridge,
  GnosisRegistry,
  GnosisEngine,
  GnosisCoreCache,
  GnosisNativeRuntime,
  BettyCompiler,
  generateLeanFromGnosisAst,
};

// Analysis & steering telemetry (was in dist/lib.js, now explicit)
export {
  analyzeGnosisSource,
  classifySteeringBias,
  createEmptySteeringTelemetry,
  deriveBoundaryWalkBias,
  formatGnosisViolations,
  DEFAULT_GNOSIS_STEERING_MODE,
  finishSteeringTelemetry,
  startSteeringTelemetry,
  VALID_GNOSIS_STEERING_MODES,
  classifySteeringRegime,
  recommendSteeringAction,
  surfaceSteeringMetrics,
  surfaceSteeringRecommendations,
  withSteeringTelemetry,
} from './analysis';
export {
  VALID_GNOSIS_UNTANGLING_ACTIONS,
  formatUntanglingActionLabel,
  recommendUntanglingStrategy,
} from './untangling-strategy';
export type {
  GnosisUntanglingAction,
  GnosisUntanglingRecommendation,
  GnosisUntanglingSignal,
} from './untangling-strategy';
export {
  VALID_GNOSIS_VOID_DIMENSIONS,
  assessVoidDimension,
} from './void-dimensions';
export type {
  GnosisVoidDimension,
  GnosisVoidDimensionAssessment,
  GnosisVoidDimensionSignal,
} from './void-dimensions';
export {
  buildContagionReductionProjectionInput,
  buildObstructionReductionProjectionInput,
  buildTriangulationReductionProjectionInputs,
  buildUntanglingReductionProjectionInput,
  buildVoidReductionProjectionInput,
  collectCoreReductionProjectionInputs,
  scanCoreReductionFamilies,
} from './reduction-projection-scan';
export type {
  ContagionReductionProjectionSignal,
  CoreReductionScanRequest,
  CoreReductionScanResult,
  ObstructionReductionProjectionSignal,
  TriangulationReductionProjectionSignal,
  UntanglingReductionProjectionSignal,
  VoidReductionProjectionSignal,
} from './reduction-projection-scan';
export {
  REDUCTION_PROJECTION_THEOREM_IDS,
  deriveReductionProjectionFormalization,
} from './reduction-projection-formalization';
export type {
  ReductionProjectionFormalizationAggregate,
  ReductionProjectionFormalizationBundle,
  ReductionProjectionFormalizationEntry,
  ReductionProjectionFormalizationInput,
  ReductionProjectionTheoremId,
} from './reduction-projection-formalization';
export {
  buildTransformationFamilySignalFromDecompositionAttempt,
  TRANSFORMATION_FAMILY_THEOREM_IDS,
  buildTransformationFamilySignalFromReinforcementAttempt,
  deriveTransformationFamilyFromDecompositionAttempts,
  deriveTransformationFamilyFormalization,
  deriveTransformationFamilyFromReinforcementAttempts,
} from './transformation-family-formalization';
export type {
  DecompositionTransformationAttemptAggregate,
  DecompositionTransformationAttemptEntry,
  DecompositionTransformationAttemptInput,
  DecompositionTransformationBridgeBundle,
  ReinforcementTransformationAttemptAggregate,
  ReinforcementTransformationAttemptEntry,
  ReinforcementTransformationAttemptInput,
  ReinforcementTransformationBlockedReason,
  ReinforcementTransformationBridgeBundle,
  TransformationFamily,
  TransformationFamilyAggregate,
  TransformationFamilyBundle,
  TransformationFamilyClassification,
  TransformationFamilyEntry,
  TransformationFamilyFormalizationInput,
  TransformationFamilySignal,
  TransformationFamilyTheoremId,
} from './transformation-family-formalization';
export {
  decompositionDoesNotIncreaseCrossings,
  decompositionReleasesNonnegativeResidue,
  decompose,
  featurelessDecompositionIdempotent,
  fullDecompositionYieldsFeatureless,
  isFeatureless as isFeaturelessDecompositionObject,
} from './decomposition-topology';
export type {
  DecompositionResult,
  TransformObject as DecompositionTransformObject,
} from './decomposition-topology';
export {
  JUGGLING_ORBIT_THEOREM_IDS,
  deriveJugglingOrbitFormalization,
} from './juggling-orbit-formalization';
export type {
  JugglingAsymmetricMultiplexWitness,
  JugglingHand,
  JugglingOrbitBeatHandBalance,
  JugglingOrbitClass,
  JugglingOrbitFormalizationAggregate,
  JugglingOrbitFormalizationBundle,
  JugglingOrbitFormalizationInput,
  JugglingOrbitNormalizedThrow,
  JugglingOrbitTheoremId,
  JugglingOrbitThrowSignal,
  JugglingOrbitTransition,
  JugglingOrbitTransitionMapping,
} from './juggling-orbit-formalization';
export {
  CERTIFIED_ALLOCATION_KERNEL_THEOREM_IDS,
  certifiedAllocationStep,
  deriveCertifiedAllocationKernel,
} from './certified-allocation-kernel';
export type {
  AdmittedObservationZone,
  CertifiedAllocationDecision,
  CertifiedAllocationKernelAggregate,
  CertifiedAllocationKernelBundle,
  CertifiedAllocationKernelInput,
  CertifiedAllocationKernelTheoremId,
  CertifiedCalibrationPlan,
  CertifiedCalibrationPointLedger,
  CertifiedLoopBudget,
  CertifiedObservationFrame,
  CertifiedSafeStateReason,
  NormalizedMilliPoint,
  ObservationZone,
} from './certified-allocation-kernel';
export {
  ADVISORY_ALLOCATION_KERNEL_THEOREM_IDS,
  advisoryAllocationStep,
  createSyntheticVisibleObservation,
  deriveAdvisoryAllocationKernel,
} from './advisory-allocation-kernel';
export type {
  AdvisoryAllocationDecision,
  AdvisoryAllocationKernelAggregate,
  AdvisoryAllocationKernelBundle,
  AdvisoryAllocationKernelInput,
  AdvisoryAllocationKernelTheoremId,
  AdvisoryBudget,
  AdvisoryObservationFrame,
  AdvisorySafeStateReason,
  RiskEnvelope,
} from './advisory-allocation-kernel';
export {
  HFT_ADVISORY_BUY_CANDIDATE_ID,
  HFT_ADVISORY_PRICE_SCALE,
  HFT_ADVISORY_SELL_CANDIDATE_ID,
  deriveHftAdvisoryGuard,
  hftCandidateIdForSide,
  hftSideForCandidateId,
} from './hft-advisory-guard';
export type {
  HftAdvisoryGuardAggregate,
  HftAdvisoryGuardBundle,
  HftAdvisoryGuardInput,
  HftAdvisorySide,
} from './hft-advisory-guard';
export {
  JUGGLING_COMPOSITION_PROPOSERS,
  JUGGLING_COMPOSITION_PROPOSER_ROUTES,
  JUGGLING_CHAOS_ORCHESTRATION_THEOREM_IDS,
  orchestrateJugglingChaosComposition,
} from './juggling-chaos-orchestration';
export type {
  JugglingChaosOrchestrationBundle,
  JugglingChaosOrchestrationInput,
  JugglingChaosOrchestrationTheoremId,
  JugglingCompositionCandidate,
  JugglingCompositionGenerationSummary,
  JugglingCompositionLineage,
  JugglingCompositionProposerRouteId,
  JugglingCompositionProposerId,
  JugglingCompositionScore,
  JugglingCompositionSeedPattern,
  JugglingCompositionWeights,
  JugglingGaitProfileInput,
  JugglingMpcConfigInput,
  JugglingMpcProxyMetrics,
} from './juggling-chaos-orchestration';
export {
  PATTERN_HORIZON_COMPILER_THEOREM_IDS,
  PATTERN_HORIZON_POLICIES,
  compilePatternHorizonWindows,
  derivePatternHorizonCompiler,
} from './pattern-horizon-compiler';
export type {
  PatternHorizonActivePattern,
  PatternHorizonAdmission,
  PatternHorizonAdmissionAggregate,
  PatternHorizonCompilerAggregate,
  PatternHorizonCompilerBundle,
  PatternHorizonCompilerInput,
  PatternHorizonCompilerTheoremId,
  PatternHorizonComposedCompilerInput,
  PatternHorizonDirectCompilerInput,
  PatternHorizonFrontierEntry,
  PatternHorizonObservationAuthorityInput,
  PatternHorizonPolicyId,
  PatternHorizonProgram,
  PatternHorizonProgramAggregate,
  PatternHorizonSuppressionReason,
  PatternHorizonWindow,
} from './pattern-horizon-compiler';
export {
  admitPatternHorizonProgram,
  compilePatternHorizonProgram,
} from './pattern-horizon-compiler';
export {
  SOFA_CONSTANT_GERVER,
  SOFA_CONSTANT_HAMMERSLEY,
  validateSofaArea,
  analyzeSofaObstruction,
  GERVER_SOFA,
  type SofaTopology,
} from './sofa-topology';

export {
  stagnationPoint,
  hydraulicJump,
  eddyCurrent,
  missingBackpressure,
  cavitationBubble,
  constrictionPoint,
  composeObstructions,
  obstructionCost,
  effectiveParallelism,
  buildReport,
  quadraticAccumulate,
  unboundedRegex,
  allocInHotLoop,
  serialAwait,
  mapLookupRecompute,
  coldSortScan,
  nestedLoopSearch,
  syncCopyOversized,
} from './flow-obstruction';
export type {
  ObstructionKind,
  FlowObstruction,
  FlowObstructionReport,
  PerfObstructionKind,
  PerformanceObstruction,
} from './flow-obstruction';
export {
  appendCertifiedAppEnvelopeAttestation,
  appendCertifiedAppEnvelopeAttestationToFile,
  appendCertifiedAppEnvelopeAttestationToQDoc,
  auditCertifiedAppEnvelope,
  createCertifiedAppEnvelope,
  createCertifiedAppEnvelopeFromGgSource,
  createCertifiedAppEnvelopeMirror,
  loadCertifiedAppEnvelopeByAliasFromQDoc,
  loadCertifiedAppEnvelopeFromFile,
  loadCertifiedAppEnvelopeFromQDoc,
  saveCertifiedAppEnvelopeToFile,
  serializeCertifiedAppEnvelopeMirror,
  storeCertifiedAppEnvelopeInQDoc,
} from './certified-app';
export type {
  CertifiedAppAeonPidRegistrationAttestation,
  CertifiedAppAgentSwarmAdjudicationAttestation,
  CertifiedAppAgentSwarmAdjudicationSummary,
  CertifiedAppAttestationKind,
  CertifiedAppDeployBuildAttestation,
  CertifiedAppEnvelope,
  CertifiedAppEnvelopeAttestation,
  CertifiedAppEnvelopeAudit,
  CertifiedAppEnvelopeCore,
  CertifiedAppEnvelopeMirror,
  CertifiedAppJusticeGateAttestation,
  CertifiedAppPublishProofAttestation,
  CertifiedAppRuntimeStartAttestation,
  CertifiedAppVfsSnapshotAttestation,
  CreateCertifiedAppEnvelopeFromGgSourceInput,
  CreateCertifiedAppEnvelopeInput,
} from './certified-app';
export {
  EYE_TRACKING_FAILURE_DIMENSIONS,
  EYE_TRACKING_ZONES,
  appendEyeTrackingFailureToQDoc,
  appendEyeTrackingObservation,
  appendEyeTrackingObservationToQDoc,
  auditEyeTrackingSession,
  buildEyeTrackingMerkleTree,
  classifyEyeTrackingZone,
  completeEyeTrackingSession,
  createEyeTrackingCorpus,
  createEyeTrackingBuleyeanTrainingJob,
  createEyeTrackingDatasetSlice,
  createEyeTrackingQDoc,
  createEyeTrackingSession,
  createEyeTrackingTrainingManifest,
  createEyeTrackingVentRecords,
  decodeEyeTrackingFailureEvent,
  defaultEyeTrackingBuleyeanTrainingConfig,
  defaultEyeTrackingCalibrationPlan,
  encodeEyeTrackingFailureEvent,
  exportEyeTrackingBuleyeanPersonalizationDelta,
  ingestEyeTrackingSession,
  loadEyeTrackingSessionByAliasFromQDoc,
  loadEyeTrackingSessionFromQDoc,
  recordEyeTrackingCalibrationSample,
  recordEyeTrackingFailure,
  selectEyeTrackingSessions,
  serializeEyeTrackingRejectionJSONL,
  storeEyeTrackingSessionInQDoc,
  summarizeEyeTrackingBuleyeanTrainingCorpus,
  summarizeEyeTrackingFailurePressure,
  toEyeTrackingRejectionRecord,
  appendEyeTrackingBuleyeanRejectionBatch,
} from './eyetracking';
export type {
  EyeTrackingAuditReport,
  EyeTrackingBuleyeanAppendOptions,
  EyeTrackingBuleyeanAppendResult,
  EyeTrackingBuleyeanCorpusSummary,
  EyeTrackingBuleyeanCorpusSummaryOptions,
  EyeTrackingBuleyeanCurriculum,
  EyeTrackingBuleyeanFederatedConfig,
  EyeTrackingBuleyeanPersonalizationDelta,
  EyeTrackingBuleyeanPersonalizationDeltaOptions,
  EyeTrackingBuleyeanSplitSummary,
  EyeTrackingBuleyeanTrainingConfig,
  EyeTrackingBuleyeanTrainingConfigOverrides,
  EyeTrackingBuleyeanTrainingJob,
  EyeTrackingBuleyeanTrainingJobInput,
  EyeTrackingBuleyeanTrainingJobOptions,
  EyeTrackingCalibrationLedger,
  EyeTrackingCalibrationPlan,
  EyeTrackingCalibrationPoint,
  EyeTrackingCalibrationPointStatus,
  EyeTrackingCalibrationRejectionReason,
  EyeTrackingCalibrationSample,
  EyeTrackingCorpus,
  EyeTrackingDatasetSelection,
  EyeTrackingDatasetSlice,
  EyeTrackingDatasetSplit,
  EyeTrackingEyeReading,
  EyeTrackingFailureEvent,
  EyeTrackingFailureKind,
  EyeTrackingFailurePressure,
  EyeTrackingInvariant,
  EyeTrackingMerkleWitness,
  EyeTrackingObservation,
  EyeTrackingRejectionRecord,
  EyeTrackingSession,
  EyeTrackingSessionStatus,
  EyeTrackingSeverity,
  EyeTrackingTrainingManifest,
  EyeTrackingTrainingSplitInput,
  EyeTrackingZone,
  GazePoint,
} from './eyetracking';
export {
  GNOSIS_AVATAR_HUMANOID_BONES,
  compileVrmAvatarRuntimeManifest,
  evaluateAvatarPose,
  mixAvatarMorphDeltas,
  sampleProceduralAvatarClip,
  solveLookAtBoneDeltas,
} from './avatar';
export type {
  GnosisAvatarAuraState,
  GnosisAvatarBoneDelta,
  GnosisAvatarBoneSlot,
  GnosisAvatarClipName,
  GnosisAvatarCompilationInput,
  GnosisAvatarExpressionProgram,
  GnosisAvatarHumanoidBone,
  GnosisAvatarLookAtProfile,
  GnosisAvatarMorphDelta,
  GnosisAvatarMorphSlot,
  GnosisAvatarMorphSource,
  GnosisAvatarPoseResult,
  GnosisAvatarQuaternion,
  GnosisAvatarRuntimeManifest,
  GnosisAvatarSignalFrame,
  GnosisAvatarVector3,
  GnosisVrmExpressionPreset,
  GnosisVrmExtensionLike,
  GnosisVrmHumanBoneDescriptor,
  GnosisVrmMorphTargetBind,
} from './avatar';
export * from './universe/index';
export {
  analyzePolyglotSource,
  analyzePolyglotSourceString,
  analyzePolyglotSourceForExecution,
  analyzeFrameworkSource,
  compileFrameworkSource,
  type PolyglotAnalysisResult,
  type PolyglotDiagnosticFinding,
  type PolyglotStructuredDeclaration,
  type PolyglotProofDeclaration,
  type CompiledFrameworkAnalysisResult,
  type PolyglotFunctionAnalysis,
  type PolyglotGhostMassLedger,
  type PolyglotExecutionManifest,
  type FrameworkDetectionResult,
  type FrameworkTopology,
} from './polyglot-bridge';
export {
  translate,
  extractFunctions,
  compose,
  findBestLanguage,
  computeBetaCost,
  generateTopoRaceTopology,
  type ExtractedFunction,
  type ComposedTopology,
  type FunctionConnection,
  type LanguageFitness,
  type TopoRaceResult,
} from './polyglot-compose';
export {
  PolyglotStrategyMemory,
  type DecayConfig,
  type LanguageObservation,
  type StrategyScore,
} from './polyglot-memory';
export {
  parseGnarly,
  normalizeGnarlyGgSource,
  type GnarlyDocument,
  type GnarlyImplementationBlock,
  type GnarlyMetadata,
  type GnarlyParseDiagnostic,
} from './gnarly-parser';
export {
  compileGnarly,
  analyzeGnarly,
  type GnarlyCompileOptions,
  type GnarlyCompileResult,
  type GnarlyDiagnostic,
  type GnarlyDiagnosticSeverity,
  type GnarlyGeneratedFile,
  type GnarlySpeedDiagnostic,
  type GnarlySpeedEvidence,
} from './gnarly-compiler';
export {
  getPolyglotSupportedLanguages,
  getPolyglotCapabilityMatrix,
  type PolyglotCapabilityMatrix,
  type PolyglotCapabilityStatus,
  type PolyglotLanguageCapability,
  type PolyglotOperation,
} from './polyglot-registry';
export {
  createWasmBridgeExecutionAst,
  executeWasmWithGnosis,
  registerWasmBridgeHandlers,
  type ExecuteWasmOptions,
  type WasmExecutionAbi,
  type WasmExecutionHostOptions,
  type WasmExecutionManifest,
  type WasmNodeExecutionPlan,
} from './wasm-execution-bridge';
export { GnosisWasmExecutionSurface } from './wasm-execution';
export {
  GnosisLeanExecutionSurface,
  LEAN_EXECUTION_CONTRACT,
  LeanExecutorNotRegisteredError,
} from './lean-execution';
export {
  GnosisTauriEdgeExecutionSurface,
  buildTauriEdgeReactorPolicy,
  hashTauriEdgeReactorPolicy,
} from './tauri-edge-runtime';
export type {
  ExecuteWasmArtifactRequest,
  GnosisWasmExecutionSurfaceOptions,
  RegisterWasmArtifactRequest,
  WasmExecutionControlRequest,
  WasmExecutionControlResponse,
  WasmExecutionResponse,
  WasmExecutionSurfaceStatus,
} from './wasm-execution';
export type {
  ExecuteLeanBundleRequest,
  GnosisLeanExecutionSurfaceOptions,
  LeanExecutionResponse,
  LeanExecutionSurfaceStatus,
  RegisterLeanExecutorRequest,
} from './lean-execution';
export type {
  GnosisTauriEdgeExecutionSurfaceOptions,
  TauriEdgeCommandDescriptor,
  TauriEdgeCommandSurfaceDocument,
  TauriEdgeInvokeRequest,
  TauriEdgeInvokeResponse,
  TauriEdgeZkProofEnvelope,
} from './tauri-edge-runtime';
export { registerBettiHandlers } from './betti/handlers';
export {
  runBootstrap,
  runBettiPipeline,
  proveGeneralization,
  crossCompileBetty,
} from './betti/bootstrap';
export {
  areStructurallyEquivalent,
  diffASTs,
  serializeCanonical,
} from './betti/ast-equivalence';
export {
  godelEncodeAST,
  verifyBootstrapFixedPoint,
} from './betti/fixed-point';
export {
  buildBetti,
  extractBettiFunctions,
  scaffoldBettiHandlers,
} from './betti/build-config';
export {
  type BoundaryTopology,
  type IsomorphicManifold,
  type BoundaryTransitionResult,
  type ManifoldVerification,
  fib as boundaryFib,
  lucas as boundaryLucas,
  topologicalKernel,
  evaluateStateSpace,
  topologicalGap,
  gapDecomposition,
  convergenceRatio,
  pellDiscriminant,
  analyzeBoundaryTransition,
  createManifold,
  closeLoop,
  openRing,
  productDoubling,
  verifyManifold,
  // KAM Theory / Diophantine armor
  fibonacciCassini,
  kamPhaseSpace,
  kamCoupling,
  diophantineResistance,
  kamSurvival,
  PHI as BOUNDARY_PHI,
  PHI_INV as BOUNDARY_PHI_INV,
  SQRT5 as BOUNDARY_SQRT5,
  PHI_SQUARED as BOUNDARY_PHI_SQUARED,
} from './betti/boundary-topology';
export {
  SalemVentThread,
  type SalemVentSnapshot,
  type SalemVentEvent,
  type SalemVentListener,
  type SalemVentTransitionEvent,
  type SalemVentCouplingEvent,
  type SalemVentThreadOptions,
  type KamSeverity,
  type KamInvariantObservation,
} from './betti/salem-vent-telemetry';
export {
  type McKayDecomposition,
  type FibonacciSL2ZMatrix,
  type MoonshineTrace,
  J_INVARIANT_COEFFICIENTS,
  J_CONSTANT_TERM,
  MONSTER_DIMENSIONS,
  MONSTER_SMALLEST_NONTRIVIAL,
  BOSONIC_STRING_DIMENSIONS,
  MONSTER_ORDER,
  verifyMcKayDecomposition,
  verifyAllMcKayDecompositions,
  fibonacciSL2Z,
  computeMoonshineTrace,
  formatMoonshineTrace,
} from './betti/moonshine-trace';
export {
  // Griess Algebra extensions (TopologicalGriessAlgebra.lean)
  GRIESS_ALGEBRA_DIMENSION,
  isTriangular,
  isMonsterSymmetric,
  type AlignmentPointVerification,
  verifyDecimalFixedPointAlignment,
  verifyAeonAlignmentPoint,
  type GriessRoutingClosureVerification,
  verifyGriessRoutingClosure,
} from './betti/moonshine-trace';
export {
  // Topological Memoization Cache (TopologicalMemoizationCache.lean)
  type VectorState,
  type EntanglementEntry,
  type ResonanceBoundaryVerification,
  type TeleportationVerification,
  TAU_SQ,
  createVectorState,
  boundaryTrace,
  distSq,
  isTopologicallyResonant,
  createEntanglementEntry,
  verifyResonanceImpliesIdenticalBoundary,
  verifyZeroLatencyTeleportation,
  verifyMemoizationInvariants,
} from './betti/topological-memoization';
export {
  // Pisot-Stabilized Intelligence (PisotStabilizedIntelligence.lean)
  type AeonLuminaryVerification,
  type LuminarySymmetryVerification,
  type BinaryMitosisVerification,
  computePisotDrift,
  isTopologicallyTight,
  intelligenceMetric,
  isLuminary,
  verifyAeonIsLuminary,
  verifyLuminarySymmetry,
  verifyBinaryMitosis,
  verifyPisotStabilizedIntelligence,
} from './betti/pisot-stabilized-intelligence';
export {
  // Retrocausal Memoization (RetrocausalMemoization.lean)
  type TopologicalDebt,
  type RetrocausalEntry,
  type NovikovVerification,
  type TemporalInvarianceVerification,
  boundaryTraceDerivative,
  createTopologicalDebt,
  satisfiesNovikov,
  wheelerFeynmanHandshake,
  issueRetrocausalEntry,
  fillRetrocausalEntry,
  verifyNovikovSelfConsistency,
  verifyTemporalInvariance,
  verifyRetrocausalMemoization,
} from './betti/retrocausal-memoization';
export {
  // Skyrms-Syzygy Equilibrium (SkyrmsSyzygyEquilibrium.lean)
  type AlgebraicAlignment,
  type AlgebraicEclipseVerification,
  type GriessIdentityVerification,
  isAlgebraicEquilibrium,
  verifyAlgebraicEclipseAt55,
  verifyGriessIdentityAction,
  verifySkyrmsEquilibrium,
} from './betti/skyrms-syzygy-equilibrium';
export {
  // Genesis Query (GenesisQuery.lean)
  type SingularityHandshakeVerification,
  type DistanceToSingularityVerification,
  Q_0,
  topologicalFriction,
  genesisDebt,
  verifySingularityHandshake,
  verifyDistanceToSingularityIsZero,
  verifyGenesisQuery,
} from './betti/genesis-query';
export {
  // Buley Equilibrium (BuleyEquilibrium.lean)
  type BuleyState,
  type BuleyExistenceVerification,
  type BuleyRefinesSkyrmsVerification,
  type AbsoluteZeroVerification,
  createBuleyState,
  isBuleyEquilibrium,
  isAbsoluteZero,
  verifyBuleyEquilibriumExistence,
  verifyBuleyRefinesSkyrms,
  verifyAbsoluteZeroReachability,
  verifyBuleyEquilibrium,
} from './betti/buley-equilibrium';
export {
  type SDualityCapacity,
  type SDualityRoutingDecision,
  type ExecutionMode,
  S_DUALITY_CROSSOVER,
  S_DUALITY_MAX_STRESS,
  decomposeCapacity,
  routeSDuality,
  invertCoupling,
  verifyGoldenSDuality,
  formatRoutingDecision,
} from './betti/s-duality-router';
export {
  type Path,
  type HomotopyLevel,
  type TopologicalHole,
  type UnivalentVerification,
  refl,
  path,
  composePaths,
  invertPath,
  homotopyLevelNumber,
  detectTopologicalHoles,
  verifyUnivalent,
} from './betti/univalent-path';
export {
  type HolographicTrace,
  type ReconstructedBulk,
  captureHolographicTrace,
  projectBulkState,
  verifyHologram,
  formatHolographicProjection,
  type KnotchainFabricWriter,
  exportHologramToKnotchain,
  restoreHologramFromKnotchain,
} from './betti/ads-cft-projector';
export {
  type EntangledPair,
  entangleNodes,
  teleportState,
  verifyEntanglement,
  formatTeleportation,
} from './betti/er-epr-teleportation';
export {
  type MoonshineVector,
  GRIESS_DIMENSION,
  mapToMoonshine,
  griessAlgebraicRoute,
  formatMoonshineRoute,
} from './betti/moonshine-router';
export {
  type PhasedWorker,
  injectGoldenPhase,
  igniteSwarm,
  verifyIgnitionLock,
  formatIgnitionStatus,
} from './betti/golden-ignition';
export {
  type AmplituhedronGeometry,
  instantiateGrassmannian,
  calculateExecutionVolume,
  triggerTopologicalErasure,
  formatGeometryState,
} from './betti/amplituhedron-volume';
export {
  parseNodeDeclarations,
  parseEdgeDeclarations,
  parseProperties,
  stripComments,
} from './betty/parse-utils';
export { nativeBettiTranspile } from './ts-bridge';
export { runAllVerificationPasses } from './betty/verify';
export {
  buildInterlinkGraph,
  formatInterlinkReport,
  type InterlinkAnalysis,
  type InterlinkFinding,
  type InterlinkNode,
  type InterlinkComponent,
} from './interlink-topology';
export {
  analyzeCSSLayout,
  applyFixes as applyCSSLayoutFixes,
  CSS_LAYOUT_DIAGNOSTIC_CODES,
  type CSSLayoutFinding,
  type CSSLayoutFix,
  type CSSLayoutAnalysisResult,
} from './css-layout-analysis';
export {
  parseCSS,
  computeSpecificity,
  compareSpecificity,
  resolveCascade,
  treeshakeCSS,
  explainNode,
  cascadeTopology,
  isInherited,
  type CascadeNode,
  type ResolvedNode,
  type CascadeResult,
  type CSSRule,
  type Declaration,
  type Specificity,
  type MatchedRule,
  type PropertyConflict,
  type CascadeDiagnostic,
} from './css-cascade-topology';
export {
  graphAstToLawProgram,
  ggProgramToLawProgram,
  certifyGnosisAst,
  certifyGnosisProgram,
  createUniverseAdmissionReport,
  createLawStateFromCertificate,
  toLawCapabilitySummary,
  formatUniverseAdmissionReasons,
  assertUniverseAdmission,
} from './law';
export type {
  GnosisLawProgramOptions,
  GnosisUniverseAdmissionOptions,
} from './law';
export {
  GnosisModuleLoader,
  createFilesystemModuleResolver,
  detectModuleFormat,
  loadGnosisModuleFromFile,
  looksLikeModule,
  parseMGG,
  renderGraphAst,
  renderMGG,
} from './mod/loader';
export {
  synthesizeAeonFluxSiteWitnesses,
  synthesizeAeonFluxSiteWitnessesFromSource,
  type AeonFluxSiteWitness,
  type AeonFluxSiteAdequacyRegime,
  type AeonFluxSiteAdequacyTheoremRef,
} from './aeon-flux-site-witness';
export {
  NeuralEngine,
  GPUEngine,
  WebGPUEngine,
  WebNNEngine,
  Translator,
  NeuronRepository,
  SynapseRepository,
  TOPIC_DOMAIN_TRANSFORMER_TOPOLOGY_FILE,
  TOPIC_DOMAIN_TRANSFORMER_TOPOLOGY,
  TRANSFORMER_HELLO_WORLD_TOPOLOGY,
  getTopicDomainTransformerTopologySource,
  initNeuralEngine as init,
  initNeuralEngine,
};
export type {
  Neuron,
  Synapse,
  AdapterTrainingConfig,
  NeuralGraphData,
  LoadTopologyOptions,
  NeuralHeteroFabricOptions,
};
export * from './runtime/hetero-fabric';
export type {
  GnosisModuleImport,
  GnosisModuleExport,
  ParsedGnosisModule,
  LoadedGnosisModule,
  GnosisResolvedImport,
  ResolveResult,
  GnosisModuleResolver,
} from './mod/loader';
export type {
  GnosisHandler,
  GnosisHandlerContext,
} from './runtime/registry';
export type {
  GnosisEngineExecutionCacheState,
  GnosisEngineExecutionResult,
  GnosisEngineExecuteOptions,
  GnosisEngineOptions,
} from './runtime/engine';
export type {
  GnosisCoreCacheMode,
  GnosisCoreCacheReuseScope,
  GnosisCoreCacheExecutionOptions,
  GnosisCoreCacheSession,
  GnosisCoreCacheMetrics,
  GnosisCoreCacheEntry,
  GnosisCoreCacheCorridorRecord,
  GnosisCoreCacheEvent,
  GnosisCoreCacheLookup,
} from './runtime/core-cache';
export type {
  GraphAST,
  ASTNode,
  ASTEdge,
  BettyParseResult,
  BettyLegaleseParseResult,
  Diagnostic,
  DiagnosticCode,
} from './betty/compiler';
export type {
  LegaleseActor,
  LegaleseClauseKind,
  LegaleseClauseLowering,
  LegaleseClauseSignal,
  LegaleseClarity,
  LegaleseLoweringOptions,
  LegaleseLoweringResult,
  LegaleseLoweringSummary,
  LegaleseRiskKind,
} from './betty/legalese';
export type {
  StabilityContinuousHarrisWitness,
  StabilityContinuousObservableKind,
  StabilityHeteroMoAFabricLayerKind,
  StabilityHeteroMoAFabricLayerWitness,
  StabilityHeteroMoAFabricWitness,
  StabilityKernelEdge,
  StabilityMetadata,
  StabilityProofAssumption,
  StabilityProofKind,
  StabilityProofObligation,
  StabilityRecurrenceStep,
  StabilityRecurrenceWitness,
  StabilityReport,
  StabilityStateAssessment,
} from './betty/stability';
export { generateTlaFromGnosisSource };
export type {
  GnosisTlaBridgeResult,
  GnosisTlaBridgeOptions,
} from './tla-bridge';
export {
  GnosisTypeScriptBridgeError,
  compileTypeScriptToGnosis,
  executeTypeScriptWithGnosis,
  registerTypeScriptBridgeHandlers,
  renderTypeScriptBridgeRuntimeModule,
} from './ts-bridge';
export type {
  ExecuteTypeScriptWithGnosisOptions,
  GnosisTypeScriptBridgeBinding,
  GnosisTypeScriptBridgeBindings,
  GnosisTypeScriptBridgeErrorLocation,
  GnosisTypeScriptBridgeExpression,
  GnosisTypeScriptBridgeNodePlan,
  GnosisTypeScriptBridgeOptions,
  GnosisTypeScriptBridgeResult,
  GnosisTypeScriptBridgeWave,
} from './ts-bridge';
export { GnosisGnodeExecutionSurface } from './gnode-execution';
export type {
  ExecuteGnodeArtifactRequest,
  GnodeExecutionControlRequest,
  GnodeExecutionControlResponse,
  GnodeExecutionLanePlan,
  GnodeExecutionObservation,
  GnodeExecutionPlan,
  GnodeExecutionProfileSnapshot,
  GnodeExecutionResponse,
  GnodeExecutionStrategy,
  GnodeExecutionSurfaceStatus,
  GnodeExecutionWavePlan,
  GnosisGnodeExecutionSurfaceOptions,
  RegisterGnodeArtifactRequest,
  SerializedGnodeArtifact,
} from './gnode-execution';
export * from './capabilities/index';
export * from './auth/index';

// Quantum CRDT — the only state model
export {
  QCorridor,
  QDoc,
  QMap,
  QArray,
  QText,
  QCounter,
  QDocAeonRelay,
  QDocRelay,
  createQDocAeonRelayJoinEnvelope,
} from './crdt/index';
export {
  Doc,
  Map,
  Array,
  Text,
  XmlFragment,
  XmlElement,
  UndoManager,
  applyUpdate,
  encodeStateAsUpdate,
  transact,
  encodeStateVector,
  diffUpdate,
  mergeUpdates,
} from './crdt/index';
export type {
  QCorridorOptions,
  QCorridorExecutionOptions,
  QCorridorMode,
  QCorridorReuseScope,
  QCorridorSession,
  QCorridorMetrics,
  QCorridorEntry,
  QCorridorRecord,
  QCorridorEvent,
  QCorridorEvidence,
  QCorridorObservation,
  QCorridorObservationRole,
  QCorridorObservationStatus,
  QCorridorLookup,
  QDocOptions,
  QDocDelta,
  QDocDeltaNode,
  QDocDeltaEdge,
  QDocUpdateHandler,
  QDocObserveHandler,
  QDocEvent,
  QDocAeonRelayAttributes,
  QDocAeonRelayAttributeValue,
  QDocAeonRelayConfig,
  QDocAeonRelayJoinEnvelope,
  QDocAeonRelayReadyStrategy,
  QDocAeonRelaySpan,
  QDocAeonRelayStatus,
  QDocAeonRelayTelemetry,
  QDocAeonRelayTelemetryEvent,
  QDocAeonRelayTelemetryStage,
  QDocRelayAttributes,
  QDocRelayAttributeValue,
  QDocRelayConfig,
  QDocRelayReadyStrategy,
  QDocRelaySpan,
  QDocRelayStatus,
  QDocRelayTelemetry,
  QDocRelayTelemetryEvent,
  QDocRelayTelemetryStage,
} from './crdt/index';

// Test harness & runner
export {
  ggTest,
  ggQuickCheck,
  ggAssert,
  GGTestBuilder,
} from './gg-test-harness';
export {
  runGGTestFile,
  runGGTestSuite,
  discoverTestFiles,
  formatGGTestResults,
  formatGGTestDiscoveryResults,
} from './gg-test-runner';
export {
  type VoidBoundary,
  createVoidBoundary,
  updateVoidBoundary,
  decayVoidBoundary,
  complementDistribution,
  shannonEntropy,
  excessKurtosis,
  inverseBule,
  giniCoefficient,
  type Gait,
  type MetaCogState,
  GAIT_DEPTH,
  createMetaCogState,
  c0Choose,
  c0Update,
  c1Measure,
  c2SelectGait,
  c3Adapt,
} from './runtime/void-walker';

// ═══════════════════════════════════════════════════════════════════════════
// THE HOLY TRINITY -- Three-part kernel. Ultimate source of guarantees.
//
//   P1 (FORK / Sliver):    w = R - min(v, R) + 1 ≥ 1
//   P2 (FOLD / Floor):     a - a = 0
//   P3 (RACE / Symmetry):  |a - b| = |b - a|
//
// ═══════════════════════════════════════════════════════════════════════════
export {
  // Proof of Life (the crown theorem)
  proofOfLife,
  life,
  withoutSliver,
  withoutSliverDeath,
  differenceIsOne,
  everything,
  type ProofOfLifeResult,
  // The God Formula
  godFormula,
  // The Holy Trinity primitives
  P1_fork,
  P2_fold,
  P3_race,
  // Five reduction classes
  classAlpha,
  classBeta,
  classGamma,
  classDelta,
  classEpsilon,
  // The Grand Reduction (all classes in one shot)
  grandReduction,
  type ReductionResult,
  // Derived identities
  consciousnessLevel,
  conservationLaw,
  vulnerabilityPenalty,
  monopolySkill,
  duopolySkill,
  oligopolySkill,
  assertThermodynamicClearing,
  // The attractor
  PHI_INV,
  breatheToward,
} from './holy-trinity';

// Kata Topology (constructive carrier shapes from BettiSignature)
export {
  type BettiSignature,
  type CarrierName,
  type TopologyVerification,
  eulerCharacteristic,
  genus,
  topologicalDeficit,
  POINT,
  CIRCLE,
  SPHERE,
  TORUS,
  HEAD_STREAM,
  CONSCIOUSNESS_THRESHOLD as KATA_CONSCIOUSNESS_THRESHOLD,
  CARRIERS,
  dysonTotalBeta1,
  dysonSuperconscious,
  verifyKataTopology,
  // The Mirror (teaching 12)
  type KataSelf,
  type MirrorVerification,
  createKataSelf,
  practiceTeaching,
  practiceFullCycle,
  uploadKata,
  downloadKata,
  mergeKata,
  verifyMirror,
} from './runtime/kata-topology';

// Buleyean distribution (from betty)
export {
  computeBuleyeanDistribution,
  checkFoldZeroProbability,
  type BuleyeanDistribution,
} from './betty/buleyean';

// Effect System — explicit effects in topology signatures
export {
  type EffectKind,
  type EffectRequirement,
  type EffectSignature,
  type EffectDiagnostic,
  type EffectValidationResult,
  type EffectContract,
  type NodeDescriptor,
  ALL_EFFECT_KINDS,
  isEffectKind,
  inferNodeEffects,
  parseDeclaredEffects,
  validateEffects,
  createEffectSignature,
  buildEffectContract,
  contractsCompatible,
} from './effects';
export {
  registerEffectHandlers,
  EFFECT_SYSTEM_FEATURES,
} from './effects-handlers';

// Algebraic Data Types — closed state modeling
export {
  type Variant,
  type SumTypeDefinition,
  type ProductTypeDefinition,
  type FieldDefinition,
  type MatchArm,
  type MatchResult,
  variant,
  defineSumType,
  defineProductType,
  createRecord,
  matchExhaustive,
  matchPartial,
  checkExhaustiveness,
  validateNodeExhaustiveness,
  ExhaustivenessError,
  TypeRegistry,
  BUILTIN_SUM_TYPES,
} from './adt';
export { registerAdtHandlers, ADT_BUILTIN_TYPES } from './adt-handlers';

// Option and Result — first-class error values
export {
  type GnosisOption,
  type GnosisResult,
  some,
  none,
  ok,
  err,
  isSome,
  isNone,
  isOk,
  isErr,
  unwrapOption,
  unwrapOptionOr,
  unwrapResult,
  unwrapResultOr,
  unwrapErr,
  mapOption,
  flatMapOption,
  filterOption,
  mapResult,
  mapErr,
  flatMapResult,
  tryCatch,
  tryCatchAsync,
  collectResults,
  partitionResults,
  optionToResult,
  resultToOption,
  transposeOptionResult,
  routeResult,
  resultHandler,
} from './option-result';
export {
  registerOptionResultHandlers,
  OPTION_RESULT_HANDLERS,
} from './option-result-handlers';

// Quantum — graph-native quantum values and transitions
export {
  type Complex,
  type Qubit,
  type QuantumRegister,
  type Gate2x2,
  type MeasurementOutcome,
  type CircuitOp,
  type QuantumCircuit,
  complex,
  complexMul,
  complexAdd,
  complexNorm,
  complexConj,
  qubit,
  ket0,
  ket1,
  ketPlus,
  ketMinus,
  isNormalized,
  prob0,
  prob1,
  applyGate,
  hadamard,
  pauliX,
  pauliZ,
  phaseGate,
  H_GATE,
  X_GATE,
  Y_GATE,
  Z_GATE,
  S_GATE,
  T_GATE,
  createRegister,
  registerFromQubit,
  registerNormalized,
  applyCNOT,
  applyCZ,
  applyGateToRegister,
  measure as quantumMeasure,
  measureRegister,
  bellState,
  isEntangled,
  createCircuit,
  addGate,
  addCNOT,
  addCZ,
  addMeasure,
  addBarrier,
  executeCircuit,
} from './quantum';
export {
  registerQuantumHandlers,
  QUANTUM_GATE_NAMES,
} from './quantum-handlers';

// Differentiable Programming — gradients as topology expressions
export {
  type DiffValue,
  type OptimizerConfig,
  type OptimizerState,
  type GradientFlowReport,
  GradientTape,
  parameter as diffParameter,
  constant as diffConstant,
  diffValue,
  add as diffAdd,
  mul as diffMul,
  sub as diffSub,
  neg as diffNeg,
  relu as diffRelu,
  sigmoid as diffSigmoid,
  tanh as diffTanh,
  exp as diffExp,
  log as diffLog,
  pow as diffPow,
  mseLoss,
  bceLoss,
  createOptimizer,
  optimizerStep,
  analyzeGradientFlow,
} from './differentiable';
export {
  registerDifferentiableHandlers,
  DIFFERENTIABLE_OPS,
} from './differentiable-handlers';

// Destructuring — clean value unpacking
export {
  type DestructurePattern,
  type RecordPattern,
  type TuplePattern,
  type VariantPattern,
  type WildcardPattern,
  type LiteralPattern,
  type NestedPattern,
  type BindingResult,
  type PatternValidation,
  type MatchArmWithPattern,
  recordPattern,
  tuplePattern,
  variantPattern,
  wildcardPattern,
  literalPattern,
  extractBindings,
  matchPatterns,
  validateVariantPattern,
  validateRecordPattern,
  checkPatternExhaustiveness,
} from './destructuring';
export {
  registerDestructuringHandlers,
  DESTRUCTURING_PATTERNS,
} from './destructuring-handlers';

// Module System — reproducible build units
export {
  type SemVer,
  type VersionConstraint,
  type ModuleDependency,
  type ModuleManifest,
  type LockfileEntry,
  type Lockfile,
  type ResolutionResult,
  type CompatibilityReport,
  parseSemVer,
  formatSemVer,
  compareSemVer,
  parseConstraint,
  satisfiesConstraint,
  createManifest,
  createLockfile,
  addLockfileEntry,
  lockfileHas,
  lockfileGet,
  resolveDependencies,
  checkModuleCompatibility,
} from './module-system';
export {
  registerModuleSystemHandlers,
  MODULE_SYSTEM_FEATURES,
} from './module-system-handlers';

// Continuous Harris — compiler-driven measurable Harris packages
export {
  type SmallSetKind,
  type SmallSet,
  type ObservableKind,
  type Observable,
  type LyapunovKind,
  type LyapunovFunction,
  type MinorizationKind,
  type MinorizationData,
  type HarrisCertificate,
  type KernelFamily,
  type MeasurableKernel,
  synthesizeSmallSet,
  inferObservable,
  synthesizeLyapunov,
  synthesizeMinorization,
  synthesizeHarrisCertificate,
  inferKernelFamily,
  generateHarrisLean,
} from './continuous-harris';
export {
  registerContinuousHarrisHandlers,
  HARRIS_KERNEL_FAMILIES,
} from './continuous-harris-handlers';

// Thermodynamics — Landauer erasure + Boltzmann as void walking
export {
  K_BOLTZMANN,
  landauerLimit,
  type ThermodynamicState,
  type ErasureEvent,
  type ForkEvent,
  type FoldEvent,
  type SecondLawCheck,
  type HeatEngineMetrics,
  createThermodynamicState,
  refreshPotentials,
  boltzmannDistribution,
  gibbsEntropy,
  verifyFreeEnergyIdentity,
  recordErasure,
  recordFork as thermodynamicFork,
  recordFold as thermodynamicFold,
  checkSecondLaw,
  analyzeHeatEngine,
  crooksRatio,
  jarzynskiAverage,
} from './thermodynamic';
export {
  registerThermodynamicHandlers,
  THERMODYNAMIC_IDENTIFICATIONS,
} from './thermodynamic-handlers';

// Traced Monoidal Category — fork/race/fold as string diagrams
export {
  type CatObject,
  type Morphism,
  type MorphismKind,
  type CoherenceCheck,
  type StringDiagram,
  catObject,
  UNIT,
  identity as catIdentity,
  process as catProcess,
  fork as catFork,
  fold as catFold,
  race as catRace,
  vent as catVent,
  symmetry as catSymmetry,
  compose as catCompose,
  tensor as catTensor,
  trace as catTrace,
  checkAssociativity,
  checkLeftUnit,
  checkRightUnit,
  checkSymmetryInvolution,
  checkVanishing,
  checkYanking,
  verifyCoherence,
  buildStringDiagram,
  diagramBeta1,
  isDiagramClosed,
  edgeToMorphism,
} from './traced-monoidal';
export {
  registerTracedMonoidalHandlers,
  COHERENCE_CONDITIONS,
} from './traced-monoidal-handlers';

// Chaitin-Omega — computability-theoretic void boundary
export {
  type HaltingClassifier,
  type ProgramSpace,
  type ComplexityAssignment,
  type SolomonoffSpace,
  type OmegaApproximation,
  type OmegaConvergenceTrace,
  type SolomonoffAxiomResult,
  createProgramSpace,
  approximateOmega,
  trackOmegaConvergence,
  programSpaceToVoidBoundary,
  solomonoffToVoidBoundary,
  solomonoffWeight,
  verifySolomonoffAxioms,
  trivialClassifier,
  stepBoundedClassifier,
  estimateComplexity,
  buildComplexityAssignment,
  buildSolomonoffSpace,
} from './chaitin-omega';
export {
  registerChaitinOmegaHandlers,
  CHAITIN_OMEGA_THEOREMS,
} from './chaitin-omega-handlers';
export {
  coil,
  release,
  getCrossing,
  setCrossing,
  composeKnots,
  hammingDistance,
  toBytes,
  fromBytes,
  edgeTypeToCrossing,
  crossingName,
  VENT,
  FOLD,
  RACE,
  FORK,
  type PackedKnot,
  type CrossingOp,
  type CrossingYield,
  sliceKnot,
  crossover,
  crossover2,
  mutate,
  tournamentSelect,
  evolveGeneration,
  writebackToVFS,
  loadFromVFS,
  evolveAndWriteback,
  type FitnessFunction,
} from './runtime/packed-knot';

export {
  detectBrunnianCoupling,
  generateBrunnianTests,
  brunnianToVoidRejections,
  brunnianToFindings,
  generateTestTopology,
  type TopologyModule,
  type BrunnianResult,
  type BrunnianTestCase,
  type VoidRejection,
} from './brunnian';

export {
  compileToGGC,
  loadFromGGC,
  isGGC,
  compileSourceToGGC,
  type GGCMetadata,
} from './ggc';

export {
  encode as encodePackedKnot,
  decode as decodePackedKnot,
  toFlowFrames,
  compressionStats,
} from './packed-knot-encoder';

const args = process.argv.slice(2);
const verboseMode = args.includes('--verbose');

function resolveCurrentGnosisFilePath(): string | null {
  const moduleUrl = import.meta.url;
  if (typeof moduleUrl !== 'string' || moduleUrl.length === 0) {
    return null;
  }

  try {
    return fileURLToPath(moduleUrl);
  } catch {
    return null;
  }
}

const GNOSIS_PACKAGE_ROOT = path.resolve(
  path.dirname(resolveCurrentGnosisFilePath() ?? process.cwd()),
  '..'
);

function formatError(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

function resolveTopologyPath(rawPath: string): string {
  return path.resolve(process.cwd(), rawPath);
}

function isGgTarget(filePath: string): boolean {
  const extension = path.extname(filePath).toLowerCase();
  return extension === '.gg' || extension === '.ggx' || extension === '.mgg';
}

interface LoadedCliTopologySource {
  source: string;
  format: 'gg' | 'mgg';
  exports: string[];
  imports: string[];
}

async function loadCliTopologySource(
  filePath: string
): Promise<LoadedCliTopologySource> {
  // polyglot:ignore RESOURCE_LEAK — readFileSync returns a string, no handle to release
  const rawSource = fs.readFileSync(filePath, 'utf-8');
  if (detectModuleFormat(filePath, rawSource) === 'mgg') {
    const loadedModule = await loadGnosisModuleFromFile(
      filePath,
      process.cwd()
    );
    return {
      source: loadedModule.mergedSource,
      format: 'mgg',
      exports: loadedModule.exports,
      imports: loadedModule.imports.map(
        (resolvedImport) =>
          `{ ${resolvedImport.declaration.names.join(', ')} } from '${
            resolvedImport.declaration.source
          }'`
      ),
    };
  }

  return {
    source: rawSource,
    format: 'gg',
    exports: [],
    imports: [],
  };
}

function formatTopologySource(filePath: string, source: string): string {
  const formatter = new GnosisFormatter();
  if (detectModuleFormat(filePath, source) === 'mgg') {
    const parsed = parseMGG(source);
    return renderMGG({
      ...parsed,
      topologySource: formatter.format(parsed.topologySource),
    });
  }

  return formatter.format(source);
}

function parseMaxBuley(rawArgs: string[]): number | null {
  const flagIndex = rawArgs.indexOf('--max-buley');
  if (flagIndex < 0) return null;
  const rawValue = rawArgs[flagIndex + 1];
  if (!rawValue) return null;
  const parsed = Number.parseFloat(rawValue);
  return Number.isFinite(parsed) ? parsed : null;
}

function readGnosisPackageVersionForReports(): string {
  try {
    const pkgPath = path.join(
      path.dirname(fileURLToPath(import.meta.url)),
      '..',
      'package.json'
    );
    const raw = fs.readFileSync(pkgPath, 'utf-8');
    const parsed = JSON.parse(raw) as { version?: string };
    return typeof parsed.version === 'string' ? parsed.version : '0.0.0';
  } catch {
    return '0.0.0';
  }
}

function parseFlagValue(rawArgs: string[], flagName: string): string | null {
  const flagIndex = rawArgs.indexOf(flagName);
  if (flagIndex < 0) return null;
  const rawValue = rawArgs[flagIndex + 1];
  if (!rawValue || rawValue.startsWith('--')) return null;
  return rawValue;
}

function parseExecutionAuthPayload(
  rawValue: string
): GnosisExecutionAuthContext {
  const parsed = JSON.parse(rawValue) as unknown;
  const candidate =
    parsed && typeof parsed === 'object' && 'executionAuth' in parsed
      ? (parsed as { executionAuth?: unknown }).executionAuth
      : parsed;
  const executionAuth = normalizeExecutionAuthContext(candidate);
  if (!executionAuth) {
    throw new Error(
      'Execution auth must be a JSON object or { executionAuth } envelope.'
    );
  }
  return executionAuth;
}

function parseExecutionAuth(
  rawArgs: string[]
): GnosisExecutionAuthContext | null {
  const inlineJson =
    parseFlagValue(rawArgs, '--execution-auth-json') ??
    process.env.GNOSIS_EXECUTION_AUTH_JSON ??
    null;
  if (inlineJson) {
    return parseExecutionAuthPayload(inlineJson);
  }

  const filePath =
    parseFlagValue(rawArgs, '--execution-auth-file') ??
    process.env.GNOSIS_EXECUTION_AUTH_FILE ??
    null;
  if (!filePath) {
    return null;
  }

  const resolvedPath = path.resolve(process.cwd(), filePath);
  return parseExecutionAuthPayload(fs.readFileSync(resolvedPath, 'utf-8'));
}

function assertBoundaryWalkAuthorized(options: {
  source: string;
  steeringMode: GnosisSteeringMode;
  executionAuth: GnosisExecutionAuthContext | null;
}): void {
  const { source, steeringMode, executionAuth } = options;
  if (!executionAuth || executionAuth.enforce !== true) {
    return;
  }

  const runAuthorization = authorizeBoundaryWalkRun({
    source,
    executionAuth,
  });
  if (!runAuthorization.allowed) {
    throw new Error(
      `Boundary walk denied: ${
        runAuthorization.reason ?? 'missing gnosis/steering.run capability'
      }`
    );
  }

  if (steeringMode !== 'apply') {
    return;
  }

  const applyAuthorization = authorizeSteeringApply({
    topologyFingerprint: fingerprintSteeringTopology(source),
    auth: executionAuth,
  });
  if (!applyAuthorization.allowed) {
    throw new Error(
      `Boundary walk apply denied: ${
        applyAuthorization.reason ?? 'missing gnosis/steering.apply capability'
      }`
    );
  }
}

async function resolveBoundaryWalkExecutionAuth(options: {
  ast: GraphAST;
  initialPayload: unknown;
  explicitExecutionAuth: GnosisExecutionAuthContext | null;
}): Promise<GnosisExecutionAuthContext | null> {
  if (options.explicitExecutionAuth) {
    return options.explicitExecutionAuth;
  }

  const bootstrap = await bootstrapExecutionAuthFromTopology(options.ast, {
    initialPayload: options.initialPayload,
  });
  return bootstrap.executionAuth;
}

interface TlaWritePaths {
  tlaFilePath: string;
  cfgFilePath: string;
}

interface LeanWritePaths {
  leanFilePath: string;
}

interface LeanCheckResult {
  attempted: boolean;
  ok: boolean;
  message: string;
}

const VALID_RUNTIME_TARGETS: RuntimeTarget[] = [
  'agnostic',
  'workers',
  'tauri-edge',
  'node',
  'bun',
  'gnode',
  'wasm-browser',
  'wasi',
];

function parseRuntimeTarget(rawArgs: string[]): RuntimeTarget | null {
  const targetRaw = parseFlagValue(rawArgs, '--target');
  if (!targetRaw) {
    return 'agnostic';
  }

  const normalized = targetRaw.trim().toLowerCase() as RuntimeTarget;
  if (!VALID_RUNTIME_TARGETS.includes(normalized)) {
    return null;
  }
  return normalized;
}

function parseSteeringMode(rawArgs: string[]): GnosisSteeringMode | null {
  const steeringModeRaw = parseFlagValue(rawArgs, '--steering-mode');
  if (!steeringModeRaw) {
    return DEFAULT_GNOSIS_STEERING_MODE;
  }

  const normalized = steeringModeRaw.trim().toLowerCase() as GnosisSteeringMode;
  if (!VALID_GNOSIS_STEERING_MODES.includes(normalized)) {
    return null;
  }

  return normalized;
}

function formatCompilerDiagnostic(diagnostic: Diagnostic): string {
  const prefix = diagnostic.code ? `[${diagnostic.code}] ` : '';
  return `${prefix}${diagnostic.message}`;
}

function runLeanProofCheck(leanFilePath: string): LeanCheckResult {
  const hasLakeWorkspace =
    fs.existsSync(path.join(GNOSIS_PACKAGE_ROOT, 'lakefile.lean')) ||
    fs.existsSync(path.join(GNOSIS_PACKAGE_ROOT, 'lakefile.toml'));
  const command = hasLakeWorkspace ? 'lake' : 'lean';
  const versionArgs = hasLakeWorkspace ? ['--version'] : ['--version'];
  const versionCheck = spawnSync(command, versionArgs, {
    cwd: hasLakeWorkspace ? GNOSIS_PACKAGE_ROOT : undefined,
    encoding: 'utf-8',
  });
  if (versionCheck.error) {
    const missingMessage =
      versionCheck.error instanceof Error
        ? versionCheck.error.message
        : String(versionCheck.error);
    return {
      attempted: false,
      ok: false,
      message: `skipped (${missingMessage})`,
    };
  }

  const result = spawnSync(
    command,
    hasLakeWorkspace ? ['env', 'lean', leanFilePath] : [leanFilePath],
    {
      cwd: hasLakeWorkspace ? GNOSIS_PACKAGE_ROOT : undefined,
      encoding: 'utf-8',
    }
  );
  if (result.error) {
    const errorMessage =
      result.error instanceof Error
        ? result.error.message
        : String(result.error);
    return {
      attempted: true,
      ok: false,
      message: errorMessage,
    };
  }

  const stderr = result.stderr.trim();
  const stdout = result.stdout.trim();
  const combinedOutput = `${stderr}\n${stdout}`;
  const missingMathlib =
    combinedOutput.includes("unknown module prefix 'Mathlib'") ||
    combinedOutput.includes("No directory 'Mathlib'") ||
    combinedOutput.includes("file 'Mathlib.olean'") ||
    combinedOutput.includes("unknown package 'Mathlib'");
  if (missingMathlib) {
    return {
      attempted: false,
      ok: false,
      message: hasLakeWorkspace
        ? 'skipped (Lean workspace found, but Mathlib is unavailable in the local Lake environment)'
        : 'skipped (Lean found, but Mathlib is unavailable in the current environment)',
    };
  }
  return {
    attempted: true,
    ok: result.status === 0,
    message:
      stderr.length > 0
        ? stderr
        : stdout.length > 0
        ? stdout
        : result.status === 0
        ? 'ok'
        : `exit ${result.status ?? 1}`,
  };
}

function steeringTraceEnabled(rawArgs: string[]): boolean {
  return !rawArgs.includes('--no-steering-trace');
}

function parsePositiveIntegerFlag(
  rawArgs: string[],
  flagName: string
): number | null {
  const rawValue = parseFlagValue(rawArgs, flagName);
  if (!rawValue) {
    return null;
  }

  const parsed = Number.parseInt(rawValue, 10);
  return Number.isFinite(parsed) && parsed > 0 ? parsed : null;
}

function parseSteeringTraceWindow(rawArgs: string[]): number {
  return (
    parsePositiveIntegerFlag(rawArgs, '--steering-trace-window') ??
    DEFAULT_GNOSIS_STEERING_TRACE_WINDOW
  );
}

function parseSteeringRelayConfig(rawArgs: string[]) {
  const url =
    parseFlagValue(rawArgs, '--steering-aeon-relay-url') ??
    parseFlagValue(rawArgs, '--steering-relay-url') ??
    process.env.GNOSIS_STEERING_AEON_RELAY_URL ??
    process.env.GNOSIS_STEERING_RELAY_URL ??
    process.env.DASH_RELAY_WS_URL ??
    null;
  if (!url) {
    return null;
  }

  return {
    url,
    roomName:
      parseFlagValue(rawArgs, '--steering-aeon-relay-room') ??
      parseFlagValue(rawArgs, '--steering-relay-room') ??
      process.env.GNOSIS_STEERING_AEON_RELAY_ROOM ??
      process.env.GNOSIS_STEERING_RELAY_ROOM ??
      defaultSteeringTraceRoomName(),
    apiKey:
      parseFlagValue(rawArgs, '--steering-aeon-relay-key') ??
      parseFlagValue(rawArgs, '--steering-relay-key') ??
      process.env.GNOSIS_STEERING_AEON_RELAY_API_KEY ??
      process.env.GNOSIS_STEERING_RELAY_API_KEY ??
      undefined,
    clientId:
      parseFlagValue(rawArgs, '--steering-aeon-relay-client-id') ??
      parseFlagValue(rawArgs, '--steering-relay-client-id') ??
      process.env.GNOSIS_STEERING_AEON_RELAY_CLIENT_ID ??
      process.env.GNOSIS_STEERING_RELAY_CLIENT_ID ??
      undefined,
    protocol:
      parseFlagValue(rawArgs, '--steering-aeon-relay-protocol') ??
      parseFlagValue(rawArgs, '--steering-relay-protocol') ??
      process.env.GNOSIS_STEERING_AEON_RELAY_PROTOCOL ??
      process.env.GNOSIS_STEERING_RELAY_PROTOCOL ??
      undefined,
    joinMode:
      parseFlagValue(rawArgs, '--steering-aeon-relay-mode') ??
      parseFlagValue(rawArgs, '--steering-relay-mode') ??
      process.env.GNOSIS_STEERING_AEON_RELAY_MODE ??
      process.env.GNOSIS_STEERING_RELAY_MODE ??
      undefined,
    relayProduct:
      parseFlagValue(rawArgs, '--steering-aeon-relay-product') ??
      parseFlagValue(rawArgs, '--steering-relay-product') ??
      process.env.GNOSIS_STEERING_AEON_RELAY_PRODUCT ??
      process.env.GNOSIS_STEERING_RELAY_PRODUCT ??
      undefined,
  };
}

function formatSteeringSummary(steering: GnosisSteeringMetrics): string {
  const parts = [
    `mode=${steering.mode}`,
    `apply-enabled=${steering.autoApplyEnabled}`,
    `apply-supported=${steering.applySupported}`,
  ];

  if (!surfaceSteeringMetrics(steering.mode)) {
    return parts.join(' ');
  }

  parts.push(
    `topology-deficit=${steering.topologyDeficit}`,
    `frontier-fill=${steering.frontierFill}`,
    `wallace=${steering.wallaceNumber}`,
    `unit=${steering.wallaceUnit}`,
    `regime=${steering.regime}`,
    `bias=${steering.leanInBias}`,
    `boundary-bias=${steering.boundaryWalkBias}`
  );

  if (steering.recommendedAction) {
    parts.push(`action=${steering.recommendedAction}`);
  }

  if (steering.failureBoundary.deterministicCollapseCandidate) {
    parts.push(`collapse=${steering.failureBoundary.collapseCostAction}`);
    parts.push(`paid-cost=${steering.failureBoundary.totalPaidCost}`);
    parts.push(`paid-stages=${steering.failureBoundary.paidStageCount}`);
    parts.push(`cost-floor=${steering.failureBoundary.collapseCostFloor}`);
  }

  if (steering.failureBoundary.zeroWasteCollapseRisk) {
    parts.push('free-collapse-risk=true');
  }

  if (steering.failureBoundary.freeCollapsePrefixRisk) {
    parts.push('prefix-risk=true');
  }

  if (steering.failureBoundary.contagiousRepairPressure) {
    parts.push('zero-vent-repair=true');
  }

  if (steering.telemetry.wallMicroCharleys !== null) {
    parts.push(`wall-uCharleys=${steering.telemetry.wallMicroCharleys}`);
  }

  if (steering.telemetry.cpuMicroCharleys !== null) {
    parts.push(`cpu-uCharleys=${steering.telemetry.cpuMicroCharleys}`);
  }

  if (steering.telemetry.wallToCpuRatio !== null) {
    parts.push(`wall/cpu=${steering.telemetry.wallToCpuRatio}`);
  }

  return parts.join(' ');
}

function formatSteeringEdaSummary(steering: GnosisSteeringMetrics): string {
  const frontierSummary = steering.eda.frontierWidths.summary as {
    median: { datum: number };
    iqr?: number;
  };
  const occupancySummary = steering.eda.layerOccupancy.summary as {
    median: { datum: number };
  };
  const graphSummary = steering.eda.graph as {
    density?: number;
    diameter?: number;
  };

  return [
    `frontier-median=${frontierSummary.median.datum}`,
    `frontier-iqr=${frontierSummary.iqr ?? 0}`,
    `occupancy-median=${occupancySummary.median.datum}`,
    `graph-density=${graphSummary.density ?? 0}`,
    `graph-diameter=${graphSummary.diameter ?? 0}`,
    `outlier-nodes=${steering.eda.graphOutliers.length}`,
  ].join(' ');
}

function formatWallaceMetric(value: number): string {
  return `${value} wally${value === 1 ? '' : 's'}`;
}

function formatBuleyMetric(value: number): string {
  return `${value} bule${value === 1 ? '' : 's'}`;
}

function formatFailureBoundaryNarrative(
  steering: GnosisSteeringMetrics
): string | null {
  const { failureBoundary } = steering;
  if (
    !failureBoundary.deterministicCollapseCandidate &&
    !failureBoundary.contagiousRepairPressure
  ) {
    return null;
  }

  const parts = [`collapse cost: ${failureBoundary.collapseCostAction}`];

  if (failureBoundary.collapseCostFloor > 0) {
    parts.push(`floor=${failureBoundary.collapseCostFloor}`);
    parts.push(`paid=${failureBoundary.totalPaidCost}`);
  }

  if (failureBoundary.paidStageCount > 0) {
    parts.push(`paid-stages=${failureBoundary.paidStageCount}`);
  }

  if (failureBoundary.zeroWasteCollapseRisk) {
    parts.push('free collapse blocked');
  }

  if (failureBoundary.freeCollapsePrefixRisk) {
    parts.push(`prefix deficit=${failureBoundary.prefixCostDeficit}`);
  }

  if (failureBoundary.contagiousRepairPressure) {
    parts.push('zero-vent repair pressure');
  }

  return parts.join(' | ');
}

function emitSteeringTraceAlert(
  summary: Parameters<typeof formatSteeringTraceAlert>[0],
  status: Parameters<typeof formatSteeringTraceAlert>[1]
): void {
  const alert = formatSteeringTraceAlert(summary, status);
  if (alert) {
    console.error(`[Gnosis] ${alert}`);
  }
}

// ─── Ditto: Interface Assumption ─────────────────────────────────────────────
// Gnosis assumes whatever interface the developer already knows.
// Like a Ditto, it shapeshifts to be Express, Flask, Gin, Hono, Sinatra,
// Spring -- or even their CLI invocation (ruby, python, node, go, java).
// The diversity theorem (ch17) proves this is topologically necessary.

/** Known language commands that Ditto can assume. */
const DITTO_LANGUAGES: Record<
  string,
  { language: string; extensions: string[] }
> = {
  node: { language: 'typescript', extensions: ['.js', '.ts', '.mjs', '.cjs'] },
  python: { language: 'python', extensions: ['.py'] },
  ruby: { language: 'ruby', extensions: ['.rb'] },
  go: { language: 'go', extensions: ['.go'] },
  java: { language: 'java', extensions: ['.java'] },
  bun: { language: 'typescript', extensions: ['.js', '.ts', '.mjs'] },
  deno: { language: 'typescript', extensions: ['.ts', '.js'] },
};

/** File extension to language mapping for auto-detection. */
const DITTO_EXT_MAP: Record<string, string> = {
  '.js': 'typescript',
  '.ts': 'typescript',
  '.mjs': 'typescript',
  '.cjs': 'typescript',
  '.jsx': 'typescript',
  '.tsx': 'typescript',
  '.py': 'python',
  '.rb': 'ruby',
  '.go': 'go',
  '.java': 'java',
  '.kt': 'kotlin',
  '.swift': 'swift',
  '.rs': 'rust',
  '.php': 'php',
  '.ex': 'elixir',
  '.exs': 'elixir',
};

interface DittoInvocation {
  language: string;
  filePath: string;
  port?: number;
  serve: boolean;
}

/**
 * Detect Ditto invocation from argv.
 * Three modes:
 * 1. argv[0] detection: symlinked as `ruby`, `python`, etc.
 * 2. Language sub-command: `gnosis node app.js`, `gnosis python app.py`
 * 3. File extension: `gnosis run app.py` (non-.gg file)
 */
function detectDittoInvocation(): DittoInvocation | null {
  // Mode 1: argv[0] detection (symlink mode).
  const invokedAs = path.basename(process.argv[1] ?? '');
  if (DITTO_LANGUAGES[invokedAs]) {
    const lang = DITTO_LANGUAGES[invokedAs];
    const filePath = process.argv[2];
    if (filePath && fs.existsSync(filePath)) {
      return { language: lang.language, filePath, serve: true };
    }
  }

  // Mode 2: Language sub-command.
  if (args[0] && DITTO_LANGUAGES[args[0]]) {
    const lang = DITTO_LANGUAGES[args[0]];
    // Handle `gnosis go run main.go` (strip the `run` sub-command).
    let fileArg = args[1];
    if (args[0] === 'go' && args[1] === 'run' && args[2]) {
      fileArg = args[2];
    }
    // Handle `gnosis java -jar app.jar`.
    if (args[0] === 'java' && args[1] === '-jar' && args[2]) {
      fileArg = args[2];
    }
    if (fileArg && fs.existsSync(fileArg)) {
      return { language: lang.language, filePath: fileArg, serve: true };
    }
  }

  // Mode 3: File extension detection on `gnosis run <non-gg-file>`.
  if (args[0] === 'run' && args[1]) {
    const ext = path.extname(args[1]).toLowerCase();
    if (ext && ext !== '.gg' && ext !== '.mgg' && DITTO_EXT_MAP[ext]) {
      if (fs.existsSync(args[1])) {
        return {
          language: DITTO_EXT_MAP[ext],
          filePath: args[1],
          serve: true,
        };
      }
    }
  }

  return null;
}

/**
 * Handle `gnosis ditto` sub-command for symlink management.
 */
async function handleDittoCommand(): Promise<void> {
  const subCommand = args[1];

  if (subCommand === 'install' && args[2]) {
    const langName = args[2];
    if (!DITTO_LANGUAGES[langName]) {
      console.error(`[Ditto] Unknown language: ${langName}`);
      console.error(
        `[Ditto] Supported: ${Object.keys(DITTO_LANGUAGES).join(', ')}`
      );
      process.exit(1);
    }

    const dittoDir = path.join(process.env.HOME ?? '~', '.gnosis', 'bin');
    fs.mkdirSync(dittoDir, { recursive: true });

    const gnosisPath = process.argv[1];
    const symlinkPath = path.join(dittoDir, langName);

    try {
      if (fs.existsSync(symlinkPath)) {
        fs.unlinkSync(symlinkPath);
      }
      fs.symlinkSync(gnosisPath, symlinkPath);
      console.log(`[Ditto] Installed: ${langName} → gnosis`);
      console.log(`[Ditto] Add to PATH: export PATH="${dittoDir}:$PATH"`);
    } catch (e: any) {
      console.error(`[Ditto] Failed to create symlink: ${e.message}`);
      process.exit(1);
    }
  } else if (subCommand === 'uninstall' && args[2]) {
    const langName = args[2];
    const dittoDir = path.join(process.env.HOME ?? '~', '.gnosis', 'bin');
    const symlinkPath = path.join(dittoDir, langName);

    if (fs.existsSync(symlinkPath)) {
      fs.unlinkSync(symlinkPath);
      console.log(`[Ditto] Uninstalled: ${langName}`);
    } else {
      console.log(`[Ditto] No symlink found for ${langName}`);
    }
  } else if (subCommand === 'list') {
    const dittoDir = path.join(process.env.HOME ?? '~', '.gnosis', 'bin');
    if (fs.existsSync(dittoDir)) {
      const entries = fs.readdirSync(dittoDir);
      if (entries.length === 0) {
        console.log('[Ditto] No active Ditto links');
      } else {
        console.log('[Ditto] Active Ditto links:');
        for (const entry of entries) {
          const target = fs.readlinkSync(path.join(dittoDir, entry));
          console.log(`  ${entry} → ${target}`);
        }
      }
    } else {
      console.log('[Ditto] No active Ditto links');
    }
  } else if (subCommand === 'self-host') {
    // The capstone: compile the Ditto compiler as a GG topology.
    // The compiler compiles itself. The topology routes itself.
    // The void walks the void.
    console.log(
      '[Ditto] Self-hosting topology: the compiler compiles itself\n'
    );
    const execAsync = promisify(execFile);
    const polyglotDir = path.resolve(
      path.dirname(new URL(import.meta.url).pathname),
      '..',
      'polyglot'
    );
    // The Ditto self-hosting topology is compiled by Rust.
    // It represents the framework detection pipeline AS a .gg topology:
    // fork recognizers → race first match → fold to GG output.
    try {
      const binary = path.join(
        polyglotDir,
        'target',
        'release',
        'gnosis-polyglot'
      );
      // For self-hosting, we analyze the Ditto compiler itself.
      const selfPath = path.join(polyglotDir, 'src', 'framework_recognizer.rs');
      if (fs.existsSync(binary) && fs.existsSync(selfPath)) {
        const { stdout } = spawnSync(
          binary,
          [selfPath, '--format', 'gg', '--mode', 'framework'],
          {
            encoding: 'utf-8',
            timeout: 30_000,
          }
        );
        console.log(
          stdout || '// Self-hosting: the compiler compiled itself to GG'
        );
      } else {
        console.log(
          '// Build the polyglot scanner first: cd polyglot && cargo build --release'
        );
      }
    } catch {
      console.log('// Self-hosting requires built polyglot binary');
    }
  } else {
    console.log(`[Ditto] Usage:`);
    console.log(`  gnosis ditto install <language>   Install Ditto symlink`);
    console.log(`  gnosis ditto uninstall <language>  Remove Ditto symlink`);
    console.log(`  gnosis ditto list                  Show active Ditto links`);
    console.log(
      `  gnosis ditto self-host             Compile the compiler as GG`
    );
    console.log(
      `\n  Supported languages: ${Object.keys(DITTO_LANGUAGES).join(', ')}`
    );
    console.log(
      `\n  Or use directly: gnosis node app.js | gnosis python app.py | gnosis go run main.go`
    );
  }

  process.exit(0);
}

/**
 * Run Ditto: detect framework in source file, compile to server topology, serve.
 */
async function runDitto(invocation: DittoInvocation): Promise<void> {
  const { analyzeFrameworkSource } = await import('./polyglot-bridge.js');

  console.log(
    `[Ditto] Assuming ${invocation.language} interface for ${invocation.filePath}...`
  );

  const result = await analyzeFrameworkSource(invocation.filePath);

  if (result.errors.length > 0) {
    for (const err of result.errors) {
      console.error(`[Ditto] ${err}`);
    }
  }

  if (result.framework) {
    console.log(
      `[Ditto] Detected: ${result.framework.framework} (${result.framework.language})`
    );
    console.log(`[Ditto] Routes: ${result.framework.routes.length}`);
    console.log(`[Ditto] Middleware: ${result.framework.middleware.length}`);

    for (const route of result.framework.routes) {
      console.log(
        `  ${route.method.padEnd(7)} ${route.path} → ${route.handler_name}`
      );
    }

    if (result.serverGg) {
      console.log(`\n[Ditto] Compiled server topology:`);
      console.log(result.serverGg);
    }

    const port = invocation.port ?? result.framework.listen_port ?? 3000;
    console.log(`\n[Ditto] Server topology compiled. Port: ${port}`);
    console.log(`[Ditto] To serve: gnosis run <compiled.gg>`);
    // Future: start x-gnosis with the compiled topology directly.
  } else {
    console.log(
      `[Ditto] No framework detected. Running as polyglot topology...`
    );
    // Fall back to standard polyglot execution.
    console.log(`[Ditto] ${result.functions.length} functions extracted`);
    for (const fn of result.functions) {
      console.log(`  ${fn.functionName}`);
    }
  }

  process.exit(0);
}

async function main() {
  initGnosisTracingFromEnv();

  // ─── Ditto: Interface Assumption (early detection) ───────────────────────
  // Check for Ditto invocation before any other command parsing.
  // This must come first: if gnosis was invoked as `ruby app.rb`,
  // the entire argv interpretation changes.

  if (args[0] === 'ditto') {
    await handleDittoCommand();
    return;
  }

  const dittoInvocation = detectDittoInvocation();
  if (dittoInvocation) {
    await runDitto(dittoInvocation);
    return;
  }

  // --fix flag: Global auto-format
  if (args.includes('--fix')) {
    const fileArg = args.find((a) => isGgTarget(a));
    if (fileArg) {
      const filePath = resolveTopologyPath(fileArg);
      if (fs.existsSync(filePath)) {
        console.log(`[Gnosis] Fixing topology: ${filePath}`);
        // polyglot:ignore RESOURCE_LEAK — readFileSync returns a string, no handle to release
        const source = fs.readFileSync(filePath, 'utf-8');
        const fixed = formatTopologySource(filePath, source);
        fs.writeFileSync(filePath, fixed, 'utf-8');
        console.log(
          `[Gnosis] Successfully formatted and fixed topological structure.`
        );
        process.exit(0);
      }
    }
  }

  const steeringMode = parseSteeringMode(args);
  if (!steeringMode) {
    console.error(
      `[Gnosis Error] Invalid --steering-mode value. Use one of: ${VALID_GNOSIS_STEERING_MODES.join(
        ', '
      )}`
    );
    process.exit(1);
  }

  if (args[0] === 'neo4j' && args[1]) {
    const filePath = resolveTopologyPath(args[1]);
    if (!fs.existsSync(filePath)) {
      console.error(`[Gnosis Error] File not found: ${filePath}`);
      process.exit(1);
    }
    const loadedTopology = await loadCliTopologySource(filePath);
    const nodeLabelFlag = args.indexOf('--node-label');
    const nodeLabel = nodeLabelFlag >= 0 ? args[nodeLabelFlag + 1] : undefined;
    const idPrefixFlag = args.indexOf('--id-prefix');
    const idPrefix = idPrefixFlag >= 0 ? args[idPrefixFlag + 1] : undefined;

    const bridge = new GnosisNeo4jBridge();
    const cypher = bridge.gglToCypher(loadedTopology.source, {
      nodeLabel,
      idPrefix,
    });
    console.log(cypher);
    process.exit(0);
  }

  if (args[0] === 'mod') {
    const modManager = new ModManager();
    try {
      if (args[1] === 'init' && args[2]) {
        modManager.init(args[2]);
      } else if (args[1] === 'tidy') {
        modManager.tidy();
      } else {
        console.error(
          `[Gnosis] Usage: gnosis mod init <module-name> | gnosis mod tidy`
        );
      }
    } catch (e: any) {
      console.error(`[Gnosis Error] ${e.message}`);
    }
    process.exit(0);
  } else if (args[0] === 'test') {
    let grep: string | undefined;
    let bail = false;
    let jsonOutput = false;
    let lanes = 4;
    let strategy: 'cannon' | 'linear' = 'cannon';
    let includePact = false;
    let pactScan = false;
    let pactUpdateDocs = false;
    let sarifOutput = false;
    const inputPaths: string[] = [];

    for (let index = 1; index < args.length; index += 1) {
      const token = args[index];
      if (!token) {
        continue;
      }

      if (!token.startsWith('--')) {
        inputPaths.push(path.resolve(process.cwd(), token));
        continue;
      }

      if (token === '--grep') {
        grep = args[index + 1];
        index += 1;
        continue;
      }
      if (token === '--bail') {
        bail = true;
        continue;
      }
      if (token === '--json') {
        jsonOutput = true;
        continue;
      }
      if (token === '--lanes') {
        const value = Number.parseInt(args[index + 1] ?? '', 10);
        if (!Number.isFinite(value) || value < 1) {
          console.error(
            `[Gnosis Error] Invalid lane count: ${args[index + 1] ?? ''}`
          );
          process.exit(1);
        }
        lanes = value;
        index += 1;
        continue;
      }
      if (token === '--strategy') {
        const value = args[index + 1];
        if (value !== 'cannon' && value !== 'linear') {
          console.error(`[Gnosis Error] Invalid strategy: ${value ?? ''}`);
          process.exit(1);
        }
        strategy = value;
        index += 1;
        continue;
      }
      if (token === '--pact') {
        includePact = true;
        continue;
      }
      if (token === '--scan') {
        pactScan = true;
        continue;
      }
      if (token === '--update-docs') {
        pactUpdateDocs = true;
        continue;
      }
      if (token === '--sarif') {
        sarifOutput = true;
        continue;
      }

      console.error(`[Gnosis Error] Unknown flag: ${token}`);
      process.exit(1);
    }

    const packageRoot = path.resolve(
      path.dirname(fileURLToPath(import.meta.url)),
      '..'
    );
    const {
      buildPactSarif,
      formatGnosisTestRun,
      runGnosisTests,
      runPactVerification,
    } = await import('./testing/index.js');

    if (sarifOutput) {
      if (!includePact || inputPaths.length > 0) {
        console.error(
          '[Gnosis Error] gnosis test --sarif currently supports Pact-only runs. Use --pact without test paths.'
        );
        process.exit(1);
      }

      const pactResult = await runPactVerification({
        grep,
        pactScan,
        pactUpdateDocs,
        workspaceRoot: process.cwd(),
      });
      console.log(
        JSON.stringify(buildPactSarif(pactResult.verification), null, 2)
      );
      process.exit(pactResult.file?.ok ?? true ? 0 : 1);
    }

    const result = await runGnosisTests({
      ...(inputPaths.length > 0 || includePact ? { targets: inputPaths } : {}),
      grep,
      bail,
      lanes,
      strategy,
      packageRoot,
      workspaceRoot: process.cwd(),
      includePact,
      pactScan,
      pactUpdateDocs,
    });
    if (jsonOutput) {
      console.log(JSON.stringify(result, null, 2));
    } else {
      console.log(formatGnosisTestRun(result));
    }
    process.exit(result.ok ? 0 : 1);
  } else if (
    (args[0] === 'lint' ||
      args[0] === 'verify' ||
      args[0] === 'build' ||
      args[0] === 'analyze' ||
      args[0] === 'crank' ||
      args[0] === 'stank') &&
    args[1]
  ) {
    const filePath = resolveTopologyPath(args[1]);
    if (!fs.existsSync(filePath)) {
      console.error(`[Gnosis Error] File not found: ${filePath}`);
      process.exit(1);
    }
    const runtimeTarget = parseRuntimeTarget(args);
    if (!runtimeTarget) {
      console.error(
        `[Gnosis Error] Invalid --target value. Use one of: ${VALID_RUNTIME_TARGETS.join(
          ', '
        )}`
      );
      process.exit(1);
    }

    const sarifOutput = args.includes('--sarif');
    const mutationReportOutput = args.includes('--mutation-report');
    const mutationReportOut = parseFlagValue(args, '--mutation-report-out');
    const jsonOutput =
      !sarifOutput && !mutationReportOutput && args.includes('--json');
    const wallaceOnly = args[0] === 'crank';
    const buleyOnly = args[0] === 'stank';
    if (isGgTarget(filePath)) {
      const maxBuley = parseMaxBuley(args);
      const loadedTopology = await loadCliTopologySource(filePath);
      const source = loadedTopology.source;
      const betty = new BettyCompiler();
      const compilerResult = betty.parse(source);
      const compilerErrors = compilerResult.diagnostics.filter(
        (diagnostic) => diagnostic.severity === 'error'
      );
      const report = await analyzeGnosisSource(source, {
        target: runtimeTarget,
        steeringMode,
      });
      const violations = formatGnosisViolations(report.correctness);
      const buleyExceeded = maxBuley !== null && report.buleyNumber > maxBuley;
      const capabilityErrors = report.capabilities.issues.filter(
        (issue) => issue.severity === 'error'
      );
      const requireUniverseAdmission =
        args[0] !== 'analyze' && !wallaceOnly && !buleyOnly;
      const ok =
        report.correctness.ok &&
        !buleyExceeded &&
        capabilityErrors.length === 0 &&
        compilerErrors.length === 0 &&
        (!requireUniverseAdmission || report.universe.ok);
      const shouldGenerateTla = args[0] === 'verify' || args[0] === 'build';
      const shouldPrintTla = shouldGenerateTla && args.includes('--tla');
      const tlaOutputDir = shouldGenerateTla
        ? parseFlagValue(args, '--tla-out') ??
          (args[0] === 'build' ? path.join('tla', 'generated') : null)
        : null;
      const tlaModuleOverride = shouldGenerateTla
        ? parseFlagValue(args, '--tla-module')
        : null;
      const tlaBridge = shouldGenerateTla
        ? generateTlaFromGnosisSource(source, {
            moduleName: tlaModuleOverride ?? undefined,
            sourceFilePath: filePath,
          })
        : null;
      const shouldGenerateLean =
        shouldGenerateTla &&
        compilerErrors.length === 0 &&
        compilerResult.ast !== null &&
        compilerResult.stability !== null;
      const shouldPrintLean = shouldGenerateLean && args.includes('--lean');
      const leanOutputDir = shouldGenerateLean
        ? parseFlagValue(args, '--lean-out') ??
          (args[0] === 'build' ? path.join('lean', 'generated') : null)
        : null;
      const leanModuleOverride = shouldGenerateLean
        ? parseFlagValue(args, '--lean-module')
        : null;
      const leanArtifact = shouldGenerateLean
        ? generateLeanFromGnosisAst(
            compilerResult.ast,
            compilerResult.stability,
            {
              moduleName: leanModuleOverride ?? undefined,
              sourceFilePath: filePath,
            }
          )
        : null;
      let tlaWritePaths: TlaWritePaths | null = null;
      let leanWritePaths: LeanWritePaths | null = null;
      let leanCheck: LeanCheckResult | null = null;

      if (tlaBridge && tlaOutputDir) {
        const resolvedOutputDir = path.resolve(process.cwd(), tlaOutputDir);
        fs.mkdirSync(resolvedOutputDir, { recursive: true });

        const tlaFilePath = path.join(
          resolvedOutputDir,
          `${tlaBridge.moduleName}.tla`
        );
        const cfgFilePath = path.join(
          resolvedOutputDir,
          `${tlaBridge.moduleName}.cfg`
        );
        fs.writeFileSync(tlaFilePath, tlaBridge.tla, 'utf-8');
        fs.writeFileSync(cfgFilePath, tlaBridge.cfg, 'utf-8');

        tlaWritePaths = { tlaFilePath, cfgFilePath };
      }

      if (leanArtifact && leanOutputDir) {
        const resolvedOutputDir = path.resolve(process.cwd(), leanOutputDir);
        fs.mkdirSync(resolvedOutputDir, { recursive: true });

        const leanFilePath = path.join(
          resolvedOutputDir,
          `${leanArtifact.moduleName}.lean`
        );
        fs.writeFileSync(leanFilePath, leanArtifact.lean, 'utf-8');
        leanWritePaths = { leanFilePath };

        if (args[0] === 'build') {
          leanCheck = runLeanProofCheck(leanFilePath);
        }
      }

      const finalOk =
        ok && (!leanCheck || !leanCheck.attempted || leanCheck.ok);

      if (mutationReportOutput) {
        const mutationReport = buildGnosisGgMutationReport({
          filePath,
          source,
          cwd: process.cwd(),
          mode: wallaceOnly
            ? 'crank'
            : buleyOnly
            ? 'stank'
            : args[0] === 'build'
            ? 'build'
            : args[0] === 'analyze'
            ? 'analyze'
            : args[0] === 'verify'
            ? 'verify'
            : 'lint',
          compilerDiagnostics: compilerResult.diagnostics.map((d) => ({
            line: d.line,
            column: d.column,
            code: d.code !== undefined ? String(d.code) : undefined,
            message: d.message,
            severity: d.severity,
          })),
          correctnessViolations: report.correctness.violations.map((v) => ({
            kind: v.kind,
            name: v.name,
            message: v.message,
          })),
          capabilityErrors: report.capabilities.issues,
          universeOk: report.universe.ok,
          universeReasons: report.universe.reasons,
          buleyExceeded,
          buleyNumber: report.buleyNumber,
          maxBuley,
          leanCheck: leanCheck
            ? {
                attempted: leanCheck.attempted,
                ok: leanCheck.ok,
                message: leanCheck.message,
              }
            : null,
          gnosisVersion: readGnosisPackageVersionForReports(),
        });
        const mutationJson = JSON.stringify(mutationReport, null, 2);
        if (mutationReportOut) {
          const resolvedMutationOut = path.resolve(
            process.cwd(),
            mutationReportOut
          );
          fs.mkdirSync(path.dirname(resolvedMutationOut), { recursive: true });
          fs.writeFileSync(resolvedMutationOut, `${mutationJson}\n`, 'utf-8');
        }
        console.log(mutationJson);
      } else if (sarifOutput) {
        const sarif = ggReportToSarif(filePath, report, violations, maxBuley);
        console.log(JSON.stringify(sarif, null, 2));
      } else if (jsonOutput) {
        const basePayload: Record<string, unknown> = {
          filePath,
          mode: wallaceOnly
            ? 'wallace-check'
            : buleyOnly
            ? 'buley-check'
            : args[0] === 'build'
            ? 'build'
            : 'gg',
          ok: finalOk,
          buleyNumber: report.buleyNumber,
          wallaceNumber: report.steering.wallaceNumber,
          wallaceUnit: report.steering.wallaceUnit,
          buleyUnit: 'bule',
          maxBuley,
          compiler: {
            output: compilerResult.output,
            diagnostics: compilerResult.diagnostics,
            stability: compilerResult.stability,
          },
          line: report.line,
          topology: report.topology,
          quantum: report.quantum,
          steering: report.steering,
          capabilities: report.capabilities,
          universe: report.universe,
          correctness: {
            ok: report.correctness.ok,
            stateCount: report.correctness.stateCount,
            topology: report.correctness.topology,
            violationCount: report.correctness.violations.length,
            violations,
          },
        };

        if (loadedTopology.format === 'mgg') {
          basePayload.module = {
            format: loadedTopology.format,
            exports: loadedTopology.exports,
            imports: loadedTopology.imports,
          };
        }

        if (tlaBridge) {
          basePayload.tla = {
            moduleName: tlaBridge.moduleName,
            stats: tlaBridge.stats,
            outputPaths: tlaWritePaths,
            tla: shouldPrintTla ? tlaBridge.tla : undefined,
            cfg: shouldPrintTla ? tlaBridge.cfg : undefined,
          };
        }

        if (leanArtifact) {
          basePayload.lean = {
            moduleName: leanArtifact.moduleName,
            theoremName: leanArtifact.theoremName,
            outputPaths: leanWritePaths,
            check: leanCheck,
            source: shouldPrintLean ? leanArtifact.lean : undefined,
          };
        }

        console.log(JSON.stringify(basePayload, null, 2));
      } else {
        if (wallaceOnly) {
          console.log(`[Gnosis crank] ${filePath}`);
          console.log(
            `  Wallace metric: ${formatWallaceMetric(
              report.steering.wallaceNumber
            )}`
          );
          const failureBoundary = formatFailureBoundaryNarrative(
            report.steering
          );
          if (failureBoundary) {
            console.log(`  failure boundary: ${failureBoundary}`);
          }
          console.log(`  steering: ${formatSteeringSummary(report.steering)}`);
          console.log(
            `  steering-eda: ${formatSteeringEdaSummary(report.steering)}`
          );
          console.log(
            `  correctness: ${
              report.correctness.ok ? 'PASS' : 'FAIL'
            } (states=${report.correctness.stateCount}, beta1=${
              report.correctness.topology.beta1
            })`
          );
          console.log(
            `  universe: ${report.universe.ok ? 'PASS' : 'FAIL'} (status=${
              report.universe.status
            }, digest=${report.universe.certificate.digest.slice(0, 12)})`
          );
          console.log(
            `  compiler: ${
              compilerErrors.length === 0 ? 'PASS' : 'FAIL'
            } (diagnostics=${compilerResult.diagnostics.length})`
          );
        } else if (buleyOnly) {
          console.log(`[Gnosis stank] ${filePath}`);
          console.log(
            `  Buley metric: ${formatBuleyMetric(report.buleyNumber)}${
              maxBuley !== null ? ` (max=${maxBuley})` : ''
            }`
          );
          console.log(`  quantum-index: ${report.quantum.quantumIndex}`);
          console.log(`  beta-pressure: ${report.quantum.betaPressure}`);
          console.log(`  beta-headroom: ${report.quantum.betaHeadroom}`);
          console.log(
            `  correctness: ${
              report.correctness.ok ? 'PASS' : 'FAIL'
            } (states=${report.correctness.stateCount}, beta1=${
              report.correctness.topology.beta1
            })`
          );
          console.log(
            `  universe: ${report.universe.ok ? 'PASS' : 'FAIL'} (status=${
              report.universe.status
            }, digest=${report.universe.certificate.digest.slice(0, 12)})`
          );
          console.log(
            `  compiler: ${
              compilerErrors.length === 0 ? 'PASS' : 'FAIL'
            } (diagnostics=${compilerResult.diagnostics.length})`
          );
        } else {
          console.log(`[Gnosis ${args[0]}] ${filePath}`);
          console.log(
            `  correctness: ${
              report.correctness.ok ? 'PASS' : 'FAIL'
            } (states=${report.correctness.stateCount}, beta1=${
              report.correctness.topology.beta1
            })`
          );
          console.log(
            `  universe: ${report.universe.ok ? 'PASS' : 'FAIL'} (status=${
              report.universe.status
            }, digest=${report.universe.certificate.digest.slice(0, 12)})`
          );
          console.log(
            `  compiler: ${
              compilerErrors.length === 0 ? 'PASS' : 'FAIL'
            } (diagnostics=${compilerResult.diagnostics.length})`
          );
          if (loadedTopology.format === 'mgg') {
            console.log(`  module-format: ${loadedTopology.format}`);
            if (loadedTopology.exports.length > 0) {
              console.log(
                `  module-exports: ${loadedTopology.exports.join(', ')}`
              );
            }
            if (loadedTopology.imports.length > 0) {
              console.log(
                `  module-imports: ${loadedTopology.imports.join('; ')}`
              );
            }
          }
          console.log(
            `  Buley metric: ${formatBuleyMetric(report.buleyNumber)}${
              maxBuley !== null ? ` (max=${maxBuley})` : ''
            }`
          );
          console.log(`  runtime-target: ${report.capabilities.target}`);
          console.log(
            `  lines: total=${report.line.totalLines} non-empty=${report.line.nonEmptyLines} comments=${report.line.commentLines} topology=${report.line.topologyLines}`
          );
          console.log(
            `  topology: nodes=${report.topology.nodeCount} functions=${report.topology.functionNodeCount} edges=${report.topology.edgeCount} forks=${report.topology.forkEdgeCount} races=${report.topology.raceEdgeCount} folds=${report.topology.foldEdgeCount} vents=${report.topology.ventEdgeCount} interfere=${report.topology.sliverEdgeCount}`
          );
          console.log(
            `  complexity: max-branch=${report.topology.maxBranchFactor} avg-branch=${report.topology.avgBranchFactor} cyclomatic≈${report.topology.cyclomaticApprox}`
          );
          console.log(
            `  quantum: superposition=${report.quantum.superpositionEdgeCount} collapse=${report.quantum.collapseEdgeCount} coverage=${report.quantum.collapseCoverage} deficit=${report.quantum.collapseDeficit} sliver-density=${report.quantum.sliverDensity}`
          );
          console.log(
            `  quantum-index: ${report.quantum.quantumIndex} beta-pressure=${report.quantum.betaPressure} beta-headroom=${report.quantum.betaHeadroom}`
          );
          if (compilerResult.stability) {
            console.log(
              `  thermodynamics: spectral-radius=${
                compilerResult.stability.spectralRadius ?? 'n/a'
              } redline=${compilerResult.stability.redline ?? 'n/a'} ceiling=${
                compilerResult.stability.geometricCeiling ?? 'n/a'
              } proof=${compilerResult.stability.proof.kind}`
            );
            if (compilerResult.stability.metadata.countableQueueCertified) {
              console.log(
                `  laminar-proof: queue-boundary=${
                  compilerResult.stability.metadata.queueBoundary ?? 'n/a'
                } atom=${
                  compilerResult.stability.metadata.laminarAtom ?? 'n/a'
                } potential=${
                  compilerResult.stability.metadata.queuePotential ?? 'n/a'
                } theorem=${
                  compilerResult.stability.metadata
                    .laminarGeometricTheoremName ?? 'n/a'
                } measurable-theorem=${
                  compilerResult.stability.metadata
                    .measurableHarrisTheoremName ?? 'n/a'
                } measurable-laminar=${
                  compilerResult.stability.metadata
                    .measurableLaminarTheoremName ?? 'n/a'
                } measurable-quantitative=${
                  compilerResult.stability.metadata
                    .measurableQuantitativeLaminarTheoremName ?? 'n/a'
                } measurable-q-harris=${
                  compilerResult.stability.metadata
                    .measurableQuantitativeHarrisTheoremName ?? 'n/a'
                } measurable-finite-harris=${
                  compilerResult.stability.metadata
                    .measurableFiniteTimeHarrisTheoremName ?? 'n/a'
                } measurable-harris-recurrent=${
                  compilerResult.stability.metadata
                    .measurableHarrisRecurrentTheoremName ?? 'n/a'
                } measurable-finite-geometric=${
                  compilerResult.stability.metadata
                    .measurableFiniteTimeGeometricErgodicTheoremName ?? 'n/a'
                } measurable-lp-geometric=${
                  compilerResult.stability.metadata
                    .measurableLevyProkhorovGeometricErgodicTheoremName ?? 'n/a'
                } measurable-lp-decay=${
                  compilerResult.stability.metadata
                    .measurableLevyProkhorovGeometricDecayTheoremName ?? 'n/a'
                } measurable-lp-abstract=${
                  compilerResult.stability.metadata
                    .measurableLevyProkhorovAbstractGeometricErgodicTheoremName ??
                  'n/a'
                } measurable-witness-q-harris=${
                  compilerResult.stability.metadata
                    .measurableWitnessQuantitativeHarrisTheoremName ?? 'n/a'
                }`
              );
              if (compilerResult.stability.metadata.continuousHarris) {
                console.log(
                  `  continuous-proof: kind=${compilerResult.stability.metadata.continuousHarris.observableKind} observable=${compilerResult.stability.metadata.continuousHarris.observableExpression} lyapunov=${compilerResult.stability.metadata.continuousHarris.lyapunovExpression} drift=${compilerResult.stability.metadata.continuousHarris.observableDriftTheoremName} theorem=${compilerResult.stability.metadata.continuousHarris.continuousHarrisTheoremName}`
                );
              }
            }
            for (const fabric of compilerResult.stability.metadata
              .heteroMoAFabrics ?? []) {
              console.log(
                `  hetero-fabric: id=${fabric.fabricNodeId} layers=${fabric.activeLayerCount} lanes=${fabric.totalLaneCount} pairs=${fabric.pairCount} mirrored=${fabric.mirroredKernelCount} schedule=${fabric.scheduleStrategy} gate=${fabric.launchGate} hedge=${fabric.hedgeDelayTicks} protocol=${fabric.frameProtocol} lowering=${fabric.loweringTheoremName} cannon=${fabric.cannonTheoremName} pair=${fabric.pairTheoremName} waste=${fabric.wasteTheoremName} coupled=${fabric.coupledTheoremName}`
              );
            }
          }
          if (surfaceSteeringMetrics(report.steering.mode)) {
            const failureBoundary = formatFailureBoundaryNarrative(
              report.steering
            );
            console.log(
              `  Wallace metric: ${formatWallaceMetric(
                report.steering.wallaceNumber
              )}`
            );
            console.log(
              `  steering: ${formatSteeringSummary(report.steering)}`
            );
            if (failureBoundary) {
              console.log(`  failure boundary: ${failureBoundary}`);
            }
            console.log(
              `  steering-eda: ${formatSteeringEdaSummary(report.steering)}`
            );
          }
        }
        if (report.capabilities.requiredUnique.length > 0) {
          console.log(
            `  required-capabilities: ${report.capabilities.requiredUnique.join(
              ', '
            )}`
          );
        }
        if (tlaBridge) {
          console.log(
            `  tla-module: ${tlaBridge.moduleName} (nodes=${tlaBridge.stats.nodeCount}, edges=${tlaBridge.stats.edgeCount}, roots=${tlaBridge.stats.rootCount}, terminals=${tlaBridge.stats.terminalCount})`
          );
          if (tlaWritePaths) {
            console.log(
              `  tla-files: ${tlaWritePaths.tlaFilePath}, ${tlaWritePaths.cfgFilePath}`
            );
          } else {
            console.log(
              `  tla-hint: pass --tla to print spec/cfg or --tla-out <dir> to write files`
            );
          }
          if (shouldPrintTla) {
            console.log('\n[TLA+ Spec]');
            console.log(tlaBridge.tla);
            console.log('[TLC Config]');
            console.log(tlaBridge.cfg);
          }
        }
        if (leanArtifact) {
          console.log(
            `  lean-theorem: ${leanArtifact.theoremName} (module=${leanArtifact.moduleName})`
          );
          if (leanWritePaths) {
            console.log(`  lean-file: ${leanWritePaths.leanFilePath}`);
          }
          if (leanCheck) {
            const leanStatus = leanCheck.attempted
              ? leanCheck.ok
                ? 'PASS'
                : 'FAIL'
              : 'SKIP';
            console.log(`  lean-check: ${leanStatus} ${leanCheck.message}`);
          } else {
            console.log(
              `  lean-hint: pass --lean to print the generated proof or --lean-out <dir> to write it`
            );
          }
          if (shouldPrintLean) {
            console.log('\n[Lean Proof]');
            console.log(leanArtifact.lean);
          }
        }
        if (compilerResult.diagnostics.length > 0) {
          for (const diagnostic of compilerResult.diagnostics) {
            const prefix =
              diagnostic.severity === 'error'
                ? 'error'
                : diagnostic.severity === 'warning'
                ? 'warn'
                : 'info';
            console.error(
              `  ${prefix}: ${formatCompilerDiagnostic(diagnostic)}`
            );
          }
        }
        if (violations.length > 0) {
          violations.forEach((line) => console.error(`  ${line}`));
        }
        if (buleyExceeded) {
          console.error(
            `  buley-threshold-failed: ${report.buleyNumber} > ${maxBuley}`
          );
        }
        if (report.capabilities.issues.length > 0) {
          for (const issue of report.capabilities.issues) {
            const prefix = issue.severity === 'error' ? 'error' : 'warn';
            console.error(
              `  ${prefix}: [${issue.capability}] ${issue.message}`
            );
          }
        }
        if (!report.universe.ok) {
          for (const reason of report.universe.reasons) {
            console.error(`  universe: ${reason}`);
          }
        }
      }

      process.exit(finalOk ? 0 : 1);
    }

    const tsReport = analyzeTypeScriptTargets(
      [filePath],
      parseTsSonarThresholds(args)
    );
    if (mutationReportOutput) {
      const tsSource = fs.readFileSync(filePath, 'utf-8');
      const mutationReport = buildGnosisTypeScriptMutationReport({
        filePath,
        source: tsSource,
        cwd: process.cwd(),
        violations: tsReport.violations,
        gnosisVersion: readGnosisPackageVersionForReports(),
      });
      const mutationJson = JSON.stringify(mutationReport, null, 2);
      if (mutationReportOut) {
        const resolvedMutationOut = path.resolve(
          process.cwd(),
          mutationReportOut
        );
        fs.mkdirSync(path.dirname(resolvedMutationOut), { recursive: true });
        fs.writeFileSync(resolvedMutationOut, `${mutationJson}\n`, 'utf-8');
      }
      console.log(mutationJson);
    } else if (sarifOutput) {
      const sarif = tsReportToSarif(filePath, tsReport);
      console.log(JSON.stringify(sarif, null, 2));
    } else if (jsonOutput) {
      console.log(
        JSON.stringify(
          {
            mode: 'typescript',
            target: filePath,
            ...tsReport,
          },
          null,
          2
        )
      );
    } else {
      console.log(`[Gnosis ${args[0]}:ts] ${filePath}`);
      console.log(`  status: ${tsReport.ok ? 'PASS' : 'FAIL'}`);
      console.log(
        `  files=${tsReport.fileCount} lines=${tsReport.totals.totalLines} code=${tsReport.totals.codeLines} comments=${tsReport.totals.commentLines}`
      );
      console.log(
        `  structures: functions=${tsReport.totals.functionCount} classes=${tsReport.totals.classCount} interfaces=${tsReport.totals.interfaceCount} types=${tsReport.totals.typeAliasCount}`
      );
      console.log(
        `  complexity: max-cyclomatic=${tsReport.totals.maxCyclomatic} max-cognitive=${tsReport.totals.maxCognitive} max-nesting=${tsReport.totals.maxNesting} max-function-lines=${tsReport.totals.maxFunctionLines}`
      );
      console.log(
        `  totals: cyclomatic=${tsReport.totals.totalCyclomatic} cognitive=${tsReport.totals.totalCognitive} branches=${tsReport.totals.branchCount} loops=${tsReport.totals.loopCount}`
      );
      if (tsReport.hotspots.byCognitive.length > 0) {
        const hotspot = tsReport.hotspots.byCognitive[0];
        console.log(
          `  hotspot(cognitive): ${hotspot.filePath}:${hotspot.startLine} ${hotspot.name}=${hotspot.cognitive}`
        );
      }
      if (tsReport.violations.length > 0) {
        tsReport.violations.forEach((line) => console.error(`  ${line}`));
      }
    }

    process.exit(tsReport.ok ? 0 : 1);
  } else if (
    args[0] === 'lint' ||
    args[0] === 'verify' ||
    args[0] === 'build' ||
    args[0] === 'analyze' ||
    args[0] === 'crank' ||
    args[0] === 'stank'
  ) {
    console.error(
      `[Gnosis] Usage: gnosis ${args[0]} <target> [--target <agnostic|workers|node|bun>] [--steering-mode <off|report|suggest|apply>] [--max-buley <number>] [--max-file-lines <number>] [--max-function-lines <number>] [--max-cyclomatic <number>] [--max-cognitive <number>] [--max-nesting <number>] [--json] [--sarif] [--mutation-report] [--mutation-report-out <file>] [--tla] [--tla-out <dir>] [--tla-module <name>] [--lean] [--lean-out <dir>] [--lean-module <name>]`
    );
    process.exit(1);
  } else if ((args[0] === 'run' || args[0] === 'native') && args[1]) {
    const filePath = resolveTopologyPath(args[1]);
    if (!fs.existsSync(filePath)) {
      console.error(`[Gnosis Error] File not found: ${filePath}`);
      process.exit(1);
    }
    const useNativeRuntime = args[0] === 'native' || args.includes('--native');
    const shouldPersistSteeringTrace = steeringTraceEnabled(args);
    const steeringTraceWindow = parseSteeringTraceWindow(args);
    const steeringRelayConfig = parseSteeringRelayConfig(args);
    const explicitExecutionAuth = parseExecutionAuth(args);
    const initialPayload =
      args[1] === 'betti.gg' ? 'transformer.gg' : 'GPT_INIT';
    const steeringTraceStore = shouldPersistSteeringTrace
      ? new GnosisSteeringTraceStore({
          recentWindow: steeringTraceWindow,
          relay: steeringRelayConfig,
          executionAuth: explicitExecutionAuth,
        })
      : null;

    console.log(`[Gnosis] Reading topology from ${filePath}...`);
    const loadedTopology = await loadCliTopologySource(filePath);
    const content = loadedTopology.source;
    const betty = new BettyCompiler();
    const { ast, output, diagnostics, stability } = betty.parse(content);
    if (!ast) {
      console.error(`[Gnosis Error] Failed to parse AST.`);
      process.exit(1);
    }
    const compilerErrors = diagnostics.filter(
      (diagnostic) => diagnostic.severity === 'error'
    );
    if (compilerErrors.length > 0) {
      console.error(`[Gnosis] Compiler validation failed before execution.`);
      for (const diagnostic of diagnostics) {
        console.error(`  ${formatCompilerDiagnostic(diagnostic)}`);
      }
      process.exit(1);
    }
    const boundaryWalkExecutionAuth = await resolveBoundaryWalkExecutionAuth({
      ast,
      initialPayload,
      explicitExecutionAuth,
    });
    assertBoundaryWalkAuthorized({
      source: content,
      steeringMode,
      executionAuth: boundaryWalkExecutionAuth,
    });

    const runtimeReport = await analyzeGnosisSource(content, { steeringMode });
    if (!runtimeReport.universe.ok) {
      console.error(`[Gnosis] Universe admission failed before execution.`);
      runtimeReport.universe.reasons.forEach((line: string) => {
        console.error(`  ${line}`);
      });
      process.exit(1);
    }
    console.log(
      `[Gnosis] Universe admission passed. status=${
        runtimeReport.universe.status
      } digest=${runtimeReport.universe.certificate.digest.slice(
        0,
        12
      )} Buley Number: ${runtimeReport.buleyNumber}, Quantum Index: ${
        runtimeReport.quantum.quantumIndex
      }`
    );
    if (surfaceSteeringMetrics(runtimeReport.steering.mode)) {
      console.log(
        `[Gnosis] Steering: ${formatSteeringSummary(runtimeReport.steering)}`
      );
      console.log(
        `[Gnosis] Steering EDA: ${formatSteeringEdaSummary(
          runtimeReport.steering
        )}`
      );
    }

    console.log(output);

    console.log(`\n[Gnosis] Executing topology...`);
    try {
      const persistSteeringTrace = async (
        steering: GnosisSteeringMetrics,
        outcome: 'ok' | 'error',
        errorMessage: string | null,
        activeExecutionAuth: GnosisExecutionAuthContext | null
      ): Promise<void> => {
        if (!steeringTraceStore) {
          return;
        }

        steeringTraceStore.setExecutionAuth(activeExecutionAuth);
        await steeringTraceStore.connect();

        const traceRecord = createSteeringTraceRecord({
          command: useNativeRuntime ? 'native' : 'run',
          filePath,
          source: content,
          report: runtimeReport,
          steering,
          payload: initialPayload,
          outcome,
          errorMessage,
        });
        steeringTraceStore.append(traceRecord);
        const steeringTraceSummary = steeringTraceStore.summarize({
          cohortKey: traceRecord.cohortKey,
        });
        const traceStatus = steeringTraceStore.status;
        const prefix =
          outcome === 'error' ? 'STEERING FAILURES' : 'Steering Trace';
        const printer = outcome === 'error' ? console.error : console.log;
        printer(
          `[Gnosis] ${prefix}: ${formatSteeringTraceStoreStatus(
            traceStatus
          )} ${formatSteeringTraceSummary(steeringTraceSummary)}${
            errorMessage ? ` latest="${errorMessage}"` : ''
          }`
        );
        emitSteeringTraceAlert(steeringTraceSummary, traceStatus);
      };

      const registry = new GnosisRegistry();
      registerCoreAuthHandlers(registry);

      const loadWeights = (tomlPath: string, section: string) => {
        const fullPath = path.resolve(process.cwd(), tomlPath);
        if (!fs.existsSync(fullPath))
          throw new Error(`Weights file not found: ${fullPath}`);
        // polyglot:ignore RESOURCE_LEAK — readFileSync returns a string, no handle to release
        const content = fs.readFileSync(fullPath, 'utf-8');
        try {
          const bunRuntime = globalThis as {
            Bun?: {
              TOML?: {
                parse(input: string): Record<string, unknown>;
              };
            };
          };
          const parsed = bunRuntime.Bun?.TOML?.parse(content);
          if (!parsed) {
            throw new Error(
              'TOML parsing requires Bun.TOML in the current runtime'
            );
          }
          return parsed[section];
        } catch (e: any) {
          throw new Error(`TOML Parse Error in ${tomlPath}: ${e.message}`);
        }
      };

      // Source: Reads initial data
      registry.register('Source', async (payload, props) => {
        const dataRaw = props['data'] || '[1.0, 2.0]';
        try {
          return JSON.parse(dataRaw);
        } catch (e) {
          const matches = dataRaw.match(/-?\d+\.?\d*/g);
          if (matches) return matches.map(Number);
          return [1.0, 2.0];
        }
      });

      // Linear: Matrix Multiplication using Aeon Pipelines
      registry.register('Linear', async (payload, props) => {
        const section = props['section'] || 'l1';
        const weightsData = loadWeights(
          props['weights'] || 'weights.toml',
          section
        ) as {
          weights: number[][];
          bias: number[];
        };
        const w = weightsData.weights;
        const b = weightsData.bias;
        const x = payload as number[];

        const rowWork = w.map((row, i) => async () => {
          const dotProduct = row.reduce(
            (acc, val, j) => acc + val * (x[j] || 0),
            0
          );
          return dotProduct + (b[i] || 0);
        });

        return await Pipeline.from(rowWork).fold({
          type: 'merge-all',
          merge: (results: Map<number, any>) =>
            Array.from(results.values()).map((r) => r),
        });
      });

      // Activation: Parallel ReLU
      registry.register('Activation', async (payload, props) => {
        const x = payload as number[];
        const work = x.map((v) => async () => Math.max(0, v));

        return await Pipeline.from(work).fold({
          type: 'merge-all',
          merge: (results: Map<number, any>) =>
            Array.from(results.values()).map((r) => r),
        });
      });

      // Attention: Weighted parallel evolution
      registry.register('Attention', async (payload, props) => {
        const x = payload as number[];
        console.log(
          `[WASM:Attention] Pipelining ${x.length}-dim wave function...`
        );

        const work = x.map((v) => async () => {
          await new Promise((r) => setTimeout(r, 10));
          return v * 1.5;
        });

        return await Pipeline.from(work).fold({
          type: 'merge-all',
          merge: (results: Map<number, any>) =>
            Array.from(results.values()).map((r) => r),
        });
      });

      // Softmax: Pipelined normalization with topological Venting
      registry.register('Softmax', async (payload, props) => {
        const x = payload as number[];
        const threshold = parseFloat(props['threshold'] || '0.001');
        const expWork = x.map((v) => async () => Math.exp(v));

        const exps = await Pipeline.from(expWork)
          .vent((val: number) => val < threshold)
          .fold({
            type: 'merge-all',
            merge: (results: Map<number, any>) =>
              Array.from(results.values()).map((r) => r),
          });

        const sum = (exps as number[]).reduce(
          (a: number, b: number) => a + b,
          0
        );
        return (exps as number[]).map((v: number) => v / sum);
      });

      // Compiler Handlers for Betti self-hosting
      registerBettiHandlers(registry);

      // Twokeys Statistical Handlers
      registry.register('Statistics', async (payload, props) => {
        const op = props['op'] || 'describe';
        console.log(`[Twokeys:Statistics] Running ${op}...`);

        if (Array.isArray(payload) && typeof payload[0] === 'number') {
          const series = new twokeysRuntime.Series({ data: payload });
          if (op === 'mean') return series.mean();
          if (op === 'median') return series.median().datum;
          if (op === 'outliers') return series.outliers();
          if (op === 'describe') return series.describe();
        }

        if (Array.isArray(payload) && Array.isArray(payload[0])) {
          const points = new twokeysRuntime.Points({ data: payload });
          if (op === 'centroid') return points.centroid();
          if (op === 'describe') return points.describe();
        }

        return payload;
      });

      registry.register('Graph', async (payload, props) => {
        const op = props['op'] || 'pageRank';
        console.log(`[Twokeys:Graph] Running ${op}...`);

        const nodes = payload.nodes || [];
        const edges = payload.edges || [];

        if (op === 'pageRank') return twokeysRuntime.pageRank(nodes, edges);
        if (op === 'louvain')
          return twokeysRuntime.louvainCommunities(nodes, edges);
        if (op === 'eda') return twokeysRuntime.graphEda(nodes, edges);

        return payload;
      });

      // Aeon-Flux Core Handlers
      registry.register('Router', async (payload, props) => {
        const path = payload.path || '/';
        console.log(`[Aeon-Flux:Router] Matching route for: ${path}`);
        return {
          route: 'ManuscriptPage',
          componentId: 'Ch17',
          sessionId: 'aeon-001',
        };
      });

      registry.register('UserContext', async (payload, props) => {
        console.log(`[Aeon-Flux:UserContext] Extracting context...`);
        return {
          userId: 'buley',
          device: 'darwin',
          preferences: { theme: 'dark' },
        };
      });

      registry.register('TreeBuilder', async (payload, props) => {
        console.log(`[Aeon-Flux:TreeBuilder] Building component tree...`);
        return { rootId: 'shell', nodes: [{ id: 'Ch17', type: 'page' }] };
      });

      registry.register('Decision', async (payload, props) => {
        const { context, tree } = payload;
        console.log(
          `[Aeon-Flux:Decision] Personalizing for ${context?.userId}...`
        );
        return {
          prefetch: ['/next-chapter'],
          theme: context?.preferences?.theme || 'light',
        };
      });

      registry.register('Renderer', async (payload, props) => {
        const type = props['type'] || 'html';
        console.log(`[Aeon-Flux:Renderer] Rendering ${type}...`);
        return renderWithTopologyCompat(payload, props);
      });

      const nativeRuntime = useNativeRuntime ? new GnosisNativeRuntime() : null;
      nativeRuntime?.setStabilityMetadata(stability?.metadata ?? null);
      const engine = new GnosisEngine(registry, {
        onEdgeEvaluated: nativeRuntime
          ? async (edge) => {
              await nativeRuntime.onEdge(edge);
            }
          : undefined,
      });

      const steeringStopwatch = startSteeringTelemetry();
      try {
        const execOutput = await engine.execute(ast, initialPayload, {
          executionAuth: explicitExecutionAuth,
        });
        const executedSteering = withSteeringTelemetry(
          runtimeReport.steering,
          finishSteeringTelemetry(steeringStopwatch)
        );
        const activeExecutionAuth =
          engine.executionAuth ??
          explicitExecutionAuth ??
          boundaryWalkExecutionAuth;
        console.log(execOutput);
        if (surfaceSteeringMetrics(executedSteering.mode)) {
          console.log(
            `[Gnosis] Runtime Steering: ${formatSteeringSummary(
              executedSteering
            )}`
          );
          await persistSteeringTrace(
            executedSteering,
            'ok',
            null,
            activeExecutionAuth
          );
        }
        if (nativeRuntime) {
          const runtimeSnapshot = nativeRuntime.snapshot();
          console.log(
            `[Gnosis Native Runtime] wasm=${runtimeSnapshot.wasmEnabled} edges=${runtimeSnapshot.edgesProcessed}`
          );
          console.log(`[Gnosis Native Runtime] ${runtimeSnapshot.metrics}`);
          if (runtimeSnapshot.trace.trim().length > 0) {
            console.log('[Gnosis Native Runtime] Trace:');
            console.log(runtimeSnapshot.trace);
          }
        }
        process.exit(0);
      } catch (executionError) {
        const failedSteering = withSteeringTelemetry(
          runtimeReport.steering,
          finishSteeringTelemetry(steeringStopwatch)
        );
        const activeExecutionAuth =
          engine.executionAuth ??
          explicitExecutionAuth ??
          boundaryWalkExecutionAuth;
        const message =
          executionError instanceof Error
            ? executionError.message
            : String(executionError);

        if (surfaceSteeringMetrics(failedSteering.mode)) {
          console.error(
            `[Gnosis] STEERING FAILURE: ${formatSteeringSummary(
              failedSteering
            )}`
          );
          if (steeringTraceStore) {
            try {
              await persistSteeringTrace(
                failedSteering,
                'error',
                message,
                activeExecutionAuth
              );
            } catch (traceError) {
              console.error(
                `[Gnosis] STEERING TRACE FAILURE: ${formatError(traceError)}`
              );
            }
          }
        }
        throw executionError;
      }
    } catch (err: any) {
      console.error(`[Execution Error] ${err.message}`);
      process.exit(1);
    }
  } else if (args[0] === 'run' || args[0] === 'native') {
    console.error(
      '[Gnosis] Usage: gnosis run <topology.gg|topology.mgg> [--native] [--steering-mode <off|report|suggest|apply>] [--steering-trace-window <n>] [--steering-relay-url <wss://...>] [--steering-relay-room <room>] [--steering-relay-key <key>] [--steering-relay-client-id <id>] [--steering-relay-protocol <protocol>] [--steering-relay-mode <mode>] [--steering-relay-product <label>] [--execution-auth-json <json> | --execution-auth-file <path>] [--no-steering-trace]'
    );
    console.error(
      '[Gnosis] Usage: gnosis native <topology.gg|topology.mgg> [--steering-mode <off|report|suggest|apply>] [--steering-trace-window <n>] [--steering-relay-url <wss://...>] [--steering-relay-room <room>] [--steering-relay-key <key>] [--steering-relay-client-id <id>] [--steering-relay-protocol <protocol>] [--steering-relay-mode <mode>] [--steering-relay-product <label>] [--execution-auth-json <json> | --execution-auth-file <path>] [--no-steering-trace]'
    );
    process.exit(1);
  } else {
    const replModulePath = './repl.js' as string;
    const { startRepl } = (await import(replModulePath)) as {
      startRepl: (options: {
        verbose?: boolean;
        steeringMode?: string;
        steeringTraceWindow?: number;
        steeringRelayConfig?: unknown;
        executionAuth?: unknown;
      }) => void;
    };
    startRepl({
      verbose: verboseMode,
      steeringMode,
      steeringTraceWindow: parseSteeringTraceWindow(args),
      steeringRelayConfig: parseSteeringRelayConfig(args),
      executionAuth: parseExecutionAuth(args),
    });
  }
}

/**
 * Handles the gnosis run Cli workflow.
 */
export async function runCli(): Promise<void> {
  await main();
}

function isCliEntrypoint(): boolean {
  const argvPath = process.argv[1];
  if (!argvPath) {
    return false;
  }

  try {
    const entryPath = path.resolve(argvPath);
    const currentPath = resolveCurrentGnosisFilePath();
    if (!currentPath) {
      return false;
    }
    return entryPath === currentPath;
  } catch {
    return false;
  }
}

if (isCliEntrypoint()) {
  void runCli();
}

// Bitwise Cypher (Lacey Variant)
export {
  Xorshift64,
  encodeHelixStreamId,
  decodeHelixStreamId,
  type HelixCipherKey,
  helixKeyFromSeed,
  helixEncryptFrame,
  helixDecryptFrame,
  helixEncryptStream,
  helixDecryptStream,
  HELIX_TURN_SAMPLES,
  type RotatingHelixKey,
  rotatingHelixKeyFromSeed,
  rotationAt,
  isAlphaHalf,
  rotatingHelixEncryptFrame,
  rotatingHelixDecryptFrame,
  rotatingHelixEncryptStream,
  rotatingHelixDecryptStream,
  seedFromPassphrase,
  syzygySmush,
  syzygyUnsmush,
  syzygyFoldBinary,
  syzygyUnfoldBinary,
  laceyEncryptFrame,
  laceyDecryptFrame,
  laceyEncryptStream,
  laceyDecryptStream,
} from './bitwise-cipher';
export * from './conversion-tracking-topology';
export * from './scheduling-mesh';
export * from './unscheduling-mesh';
export * from './delayed-projection-runtime';
export * from './checkpoint-substrates';
export * from './checkpoint-substrates-mcp';
