// @a0n/.ignored_gnosis/src/information-geometry.ts
/**
 * information-geometry.ts -- The manifold structure of void boundaries
 *
 * The identification:
 *   - VoidBoundary = point on a statistical manifold
 *   - complementDistribution = coordinate chart (probability simplex)
 *   - Fisher information = metric tensor on the manifold
 *   - KL divergence = geodesic distance
 *   - Natural gradient = steepest descent in Fisher metric
 *   - eta = curvature parameter
 *
 * The formal ledger entry: the space of VoidBoundaries maps to a statistical
 * manifold, and void walking traces a geodesic on it.
 */

import {
  type VoidBoundary,
  complementDistribution,
  shannonEntropy,
} from './void';

// ============================================================================
// Fisher information matrix
// ============================================================================

/**
 * Fisher information matrix for a complement distribution.
 * F_{ij} = E[∂log p / ∂θ_i · ∂log p / ∂θ_j]
 *
 * For the exponential family (which complement distribution is),
 * F_{ij} = δ_{ij} / p_i (diagonal).
 */
export function fisherInformation(
  boundary: VoidBoundary,
  eta: number
): number[] {
  const dist = complementDistribution(boundary, eta);
  return dist.map((p) => (p > 1e-15 ? 1 / p : 0));
}

/**
 * Fisher information determinant (product of diagonal entries).
 * Measures the "volume" of the statistical manifold at this point.
 */
export function fisherDeterminant(boundary: VoidBoundary, eta: number): number {
  const fisher = fisherInformation(boundary, eta);
  return fisher.reduce((a, b) => a * b, 1);
}

// ============================================================================
// Divergences -- distances between void boundaries
// ============================================================================

/**
 * KL divergence: D_KL(P || Q) = Σ p_i log(p_i / q_i).
 * Asymmetric. Measures information lost when Q is used to approximate P.
 */
export function klDivergence(p: number[], q: number[]): number {
  let kl = 0;
  for (let i = 0; i < p.length; i++) {
    const pi = p[i];
    const qi = q[i];
    if (pi !== undefined && qi !== undefined && pi > 1e-15 && qi > 1e-15) {
      kl += pi * Math.log(pi / qi);
    }
  }
  return kl;
}

/**
 * Symmetric KL divergence: (D_KL(P||Q) + D_KL(Q||P)) / 2.
 */
export function symmetricKL(p: number[], q: number[]): number {
  return (klDivergence(p, q) + klDivergence(q, p)) / 2;
}

/**
 * Jensen-Shannon divergence: JSD(P, Q) = H((P+Q)/2) - (H(P) + H(Q))/2.
 * Symmetric, bounded [0, ln 2], and is a true metric (square root).
 */
export function jensenShannonDivergence(p: number[], q: number[]): number {
  const m = p.map((pi, i) => {
    const qi = q[i];
    if (qi === undefined) {
      return pi / 2;
    }
    return (pi + qi) / 2;
  });
  const hm = shannonEntropy(m);
  const hp = shannonEntropy(p);
  const hq = shannonEntropy(q);
  return hm - (hp + hq) / 2;
}

/**
 * KL divergence between two void boundaries' complement distributions.
 */
export function voidDivergence(
  a: VoidBoundary,
  b: VoidBoundary,
  eta: number
): number {
  const pa = complementDistribution(a, eta);
  const pb = complementDistribution(b, eta);
  return klDivergence(pa, pb);
}

// ============================================================================
// Natural gradient -- steepest descent on the Fisher manifold
// ============================================================================

/**
 * Natural gradient: F^{-1} ∇L.
 * In the exponential family, F^{-1}_{ii} = p_i, so the natural gradient
 * is simply grad_i * p_i.
 *
 * This is the connection between void walking and natural gradient descent:
 * the complement distribution formalizes the Fisher-inverse scaling.
 */
export function naturalGradient(
  boundary: VoidBoundary,
  euclideanGradient: number[],
  eta: number
): number[] {
  const dist = complementDistribution(boundary, eta);
  return euclideanGradient.map((g, i) => {
    const di = dist[i];
    return di === undefined ? g : g * di;
  });
}

/**
 * The key identification: c0Choose with complement distribution
 * is natural gradient sampling. The walker takes steps proportional
 * to p_i, which is exactly the Fisher-inverse scaling.
 */
export function isNaturalGradientSampling(
  complementDist: number[],
  fisherInverse: number[],
  tolerance: number = 1e-10
): boolean {
  // Fisher inverse for exponential family: F^{-1}_{ii} = p_i
  // Complement distribution: p_i
  // They are the same!
  const totalFisher = fisherInverse.reduce((a, b) => a + b, 0);
  const normalizedFisher = fisherInverse.map((f) => f / totalFisher);
  return complementDist.every(
    (p, i) => {
      const normalized = normalizedFisher[i];
      return p !== undefined && normalized !== undefined
        ? Math.abs(p - normalized) < tolerance
        : false;
    }
  );
}

// ============================================================================
// Manifold curvature
// ============================================================================

/**
 * Scalar curvature of the statistical manifold at a point.
 * For the categorical distribution (simplex), the scalar curvature
 * is related to the number of dimensions and the concentration.
 *
 * High curvature = peaked distribution = exploitative walker.
 * Low curvature = flat distribution = exploratory walker.
 */
export function scalarCurvature(boundary: VoidBoundary, eta: number): number {
  const dist = complementDistribution(boundary, eta);
  const n = dist.length;
  if (n < 2) return 0;

  // For categorical distributions, curvature relates to concentration
  // Higher Herfindahl index = more concentrated = higher curvature
  const hhi = dist.reduce((s, p) => s + p * p, 0);
  // Normalize: uniform has HHI = 1/n, delta has HHI = 1
  return (hhi - 1 / n) / (1 - 1 / n);
}

/**
 * Geodesic distance between two points on the probability simplex.
 * Uses the Fisher-Rao metric: d(P, Q) = 2 arccos(Σ √(p_i q_i)).
 */
export function fisherRaoDistance(p: number[], q: number[]): number {
  let bhattacharyya = 0;
  for (let i = 0; i < p.length; i++) {
    const pi = p[i];
    const qi = q[i];
    if (pi === undefined || qi === undefined) {
      continue;
    }
    bhattacharyya += Math.sqrt(pi * qi);
  }
  // Clamp for numerical stability
  bhattacharyya = Math.min(1, Math.max(-1, bhattacharyya));
  return 2 * Math.acos(bhattacharyya);
}

/**
 * Geodesic distance between two void boundaries.
 */
export function voidGeodesicDistance(
  a: VoidBoundary,
  b: VoidBoundary,
  eta: number
): number {
  return fisherRaoDistance(
    complementDistribution(a, eta),
    complementDistribution(b, eta)
  );
}
