// @a0n/.ignored_gnosis/src/juggling-chaos-orchestration.test.ts
import { describe, it, expect } from 'bun:test';
import {
  orchestrateJugglingChaosComposition,
  JUGGLING_CHAOS_ORCHESTRATION_THEOREM_IDS,
  type JugglingChaosOrchestrationInput,
  type JugglingCompositionSeedPattern,
} from './juggling-chaos-orchestration';
import { JUGGLING_ORBIT_THEOREM_IDS } from './juggling-orbit-formalization';

const CASCADE_CORPUS: readonly JugglingCompositionSeedPattern[] = [
  {
    patternId: 'cascade-3',
    throws: [
      { beat: 0, fromHand: 'right', flightBeats: 3 },
      { beat: 1, fromHand: 'left', flightBeats: 3 },
      { beat: 2, fromHand: 'right', flightBeats: 3 },
    ],
    styleTags: ['basic', 'cascade'],
  },
  {
    patternId: 'shower-3',
    throws: [
      { beat: 0, fromHand: 'right', flightBeats: 5 },
      { beat: 1, fromHand: 'left', flightBeats: 1 },
      { beat: 2, fromHand: 'right', flightBeats: 5 },
    ],
    styleTags: ['shower'],
  },
];

const BASE_INPUT: JugglingChaosOrchestrationInput = {
  period: 3,
  corpus: CASCADE_CORPUS,
  populationSize: 8,
  generations: 3,
  throwsPerPattern: 6,
  seed: 42,
  generatedAt: '2026-04-10T00:00:00Z',
};

describe('juggling-chaos-orchestration', () => {
  it('exports exactly 4 composition theorem IDs', () => {
    expect(JUGGLING_CHAOS_ORCHESTRATION_THEOREM_IDS).toHaveLength(4);
    expect(JUGGLING_CHAOS_ORCHESTRATION_THEOREM_IDS).toContain(
      'THM-JUGGLING-COMPOSITION-CORPUS-DETERMINISTIC'
    );
    expect(JUGGLING_CHAOS_ORCHESTRATION_THEOREM_IDS).toContain(
      'THM-JUGGLING-COMPOSITION-EVOLUTION-ELITIST-MONOTONE'
    );
  });

  it('winner score is positive with a non-trivial corpus', () => {
    const result = orchestrateJugglingChaosComposition(BASE_INPUT);
    expect(result.winner.score.total).toBeGreaterThan(0);
  });

  it('winner orbit has zero sorry (all theorem IDs present)', () => {
    const result = orchestrateJugglingChaosComposition(BASE_INPUT);
    for (const id of JUGGLING_ORBIT_THEOREM_IDS) {
      expect(result.winner.orbit.theoremIds).toContain(id);
    }
  });

  it('produces the declared number of generation summaries', () => {
    const result = orchestrateJugglingChaosComposition(BASE_INPUT);
    expect(result.generationSummaries).toHaveLength(
      BASE_INPUT.generations!
    );
  });

  it('elitist monotone: best score never decreases across generations', () => {
    const result = orchestrateJugglingChaosComposition(BASE_INPUT);
    for (let i = 1; i < result.generationSummaries.length; i++) {
      const prev = result.generationSummaries[i - 1]!;
      const curr = result.generationSummaries[i]!;
      expect(curr.bestScore).toBeGreaterThanOrEqual(prev.bestScore);
    }
  });

  it('finalists are sorted by total score descending', () => {
    const result = orchestrateJugglingChaosComposition(BASE_INPUT);
    for (let i = 1; i < result.finalists.length; i++) {
      const prev = result.finalists[i - 1]!;
      const curr = result.finalists[i]!;
      expect(prev.score.total).toBeGreaterThanOrEqual(curr.score.total);
    }
  });

  it('deterministic: same seed produces identical winner', () => {
    const r1 = orchestrateJugglingChaosComposition(BASE_INPUT);
    const r2 = orchestrateJugglingChaosComposition(BASE_INPUT);
    expect(r1.winner.candidateId).toBe(r2.winner.candidateId);
    expect(r1.winner.signature).toBe(r2.winner.signature);
    expect(r1.winner.score.total).toBe(r2.winner.score.total);
  });

  it('different seeds produce different winners', () => {
    const r1 = orchestrateJugglingChaosComposition(BASE_INPUT);
    const r2 = orchestrateJugglingChaosComposition({
      ...BASE_INPUT,
      seed: 999,
    });
    // Signatures should differ (probabilistically guaranteed with different seeds)
    expect(r1.winner.signature).not.toBe(r2.winner.signature);
  });

  it('corpus digest reflects input corpus', () => {
    const result = orchestrateJugglingChaosComposition(BASE_INPUT);
    expect(result.corpusDigest.patternCount).toBe(2);
    expect(result.corpusDigest.uniqueThrowTemplateCount).toBeGreaterThan(0);
    expect(result.corpusDigest.styleTagCount).toBeGreaterThan(0);
  });

  it('MPC proxy metrics are bounded [0, 1]', () => {
    const result = orchestrateJugglingChaosComposition(BASE_INPUT);
    const mpc = result.winner.score.mpc;
    for (const value of [
      mpc.horizonPressure,
      mpc.nonConvexRisk,
      mpc.dimensionalLoad,
      mpc.uncertaintyLoad,
      mpc.warmStartGain,
    ]) {
      expect(value).toBeGreaterThanOrEqual(0);
      expect(value).toBeLessThanOrEqual(1);
    }
  });

  it('empty corpus falls back to default patterns', () => {
    const result = orchestrateJugglingChaosComposition({
      ...BASE_INPUT,
      corpus: [],
    });
    expect(result.winner).toBeDefined();
    expect(result.winner.score.total).toBeGreaterThanOrEqual(0);
    expect(result.corpusDigest.patternCount).toBeGreaterThanOrEqual(1);
  });

  it('proposer modes all produce valid output', () => {
    for (const proposer of ['markov', 'corpus-remix', 'hybrid'] as const) {
      const result = orchestrateJugglingChaosComposition({
        ...BASE_INPUT,
        proposer,
      });
      expect(result.winner).toBeDefined();
      expect(result.winner.score.total).toBeGreaterThanOrEqual(0);
      expect(result.configuration.proposer).toBe(proposer);
    }
  });

  it('dependent theorem IDs link to orbit formalization', () => {
    const result = orchestrateJugglingChaosComposition(BASE_INPUT);
    expect(result.dependentTheoremIds).toEqual(JUGGLING_ORBIT_THEOREM_IDS);
  });
});
