// @a0n/.ignored_gnosis/src/juggling-orbit-formalization.test.ts
import { describe, it, expect } from 'bun:test';
import {
  deriveJugglingOrbitFormalization,
  type JugglingOrbitFormalizationInput,
  type JugglingOrbitThrowSignal,
  JUGGLING_ORBIT_THEOREM_IDS,
} from './juggling-orbit-formalization';

describe('juggling-orbit-formalization', () => {
  // Siteswap "3" is period 2: R-throw-3, L-throw-3, alternating hands.
  // (beat + flightBeats) % period maps back correctly.
  const CASCADE_3: JugglingOrbitFormalizationInput = {
    period: 2,
    throws: [
      { beat: 0, fromHand: 'right', flightBeats: 3 },
      { beat: 1, fromHand: 'left', flightBeats: 3 },
    ],
  };

  it('exports exactly 5 theorem IDs', () => {
    expect(JUGGLING_ORBIT_THEOREM_IDS).toHaveLength(5);
    expect(JUGGLING_ORBIT_THEOREM_IDS).toContain(
      'THM-JUGGLING-ORBIT-TRANSITION-DETERMINISTIC'
    );
    expect(JUGGLING_ORBIT_THEOREM_IDS).toContain(
      'THM-JUGGLING-ORBIT-DECOMPOSITION-FOLD'
    );
  });

  it('3-ball cascade produces zero conservation mismatches', () => {
    const result = deriveJugglingOrbitFormalization(CASCADE_3);
    expect(result.aggregate.conservationMismatchCount).toBe(0);
    for (const balance of result.beatHandBalances) {
      expect(balance.balanced).toBe(true);
      expect(balance.mismatch).toBe(0);
    }
  });

  it('3-ball cascade has at least one closed orbit', () => {
    const result = deriveJugglingOrbitFormalization(CASCADE_3);
    expect(result.aggregate.closedOrbitCount).toBeGreaterThanOrEqual(1);
    expect(result.aggregate.openOrbitCount).toBe(0);
  });

  it('3-ball cascade has zero open transitions', () => {
    const result = deriveJugglingOrbitFormalization(CASCADE_3);
    expect(result.aggregate.openTransitionCount).toBe(0);
  });

  it('normalizes throws to the correct period', () => {
    const result = deriveJugglingOrbitFormalization(CASCADE_3);
    expect(result.period).toBe(2);
    for (const t of result.normalizedThrows) {
      expect(t.beat).toBeGreaterThanOrEqual(0);
      expect(t.beat).toBeLessThan(2);
    }
  });

  it('asymmetric multiplex fires on a known pattern', () => {
    // Two throws from the same beat/hand with different angles
    const multiplex: JugglingOrbitFormalizationInput = {
      period: 2,
      throws: [
        { beat: 0, fromHand: 'right', flightBeats: 2, angleDeg: 10, speed: 5 },
        { beat: 0, fromHand: 'right', flightBeats: 3, angleDeg: 45, speed: 5 },
        { beat: 1, fromHand: 'left', flightBeats: 2 },
      ],
    };
    const result = deriveJugglingOrbitFormalization(multiplex);
    expect(result.asymmetricMultiplexWitnesses.length).toBeGreaterThanOrEqual(1);
    const asymmetricWitness = result.asymmetricMultiplexWitnesses.find(
      (w) => w.asymmetric
    );
    expect(asymmetricWitness).toBeDefined();
    expect(asymmetricWitness!.outgoingCount).toBeGreaterThanOrEqual(2);
    expect(asymmetricWitness!.angleRangeDeg).toBeGreaterThan(0);
  });

  it('machine-only boundary fires with sufficient separation', () => {
    const pattern: JugglingOrbitFormalizationInput = {
      period: 2,
      throws: [
        { beat: 0, fromHand: 'right', flightBeats: 2, angleDeg: 0, speed: 3 },
        { beat: 0, fromHand: 'right', flightBeats: 3, angleDeg: 45, speed: 8 },
        { beat: 1, fromHand: 'left', flightBeats: 2 },
      ],
      machineOnlyMultiplexThreshold: 2,
      machineOnlyAngleThresholdDeg: 12,
      machineOnlySpeedRatioThreshold: 1.3,
    };
    const result = deriveJugglingOrbitFormalization(pattern);
    const machineWitness = result.asymmetricMultiplexWitnesses.find(
      (w) => w.machineOnlyCandidate
    );
    expect(machineWitness).toBeDefined();
  });

  it('single throw pattern produces open transitions', () => {
    const single: JugglingOrbitFormalizationInput = {
      period: 1,
      throws: [{ beat: 0, fromHand: 'right', flightBeats: 2 }],
    };
    const result = deriveJugglingOrbitFormalization(single);
    expect(result.aggregate.throwCount).toBe(1);
  });

  it('empty pattern produces empty results', () => {
    const empty: JugglingOrbitFormalizationInput = {
      period: 1,
      throws: [],
    };
    const result = deriveJugglingOrbitFormalization(empty);
    expect(result.aggregate.throwCount).toBe(0);
    expect(result.aggregate.orbitCount).toBe(0);
  });

  it('deterministic: same input produces identical output', () => {
    const result1 = deriveJugglingOrbitFormalization({
      ...CASCADE_3,
      generatedAt: '2026-01-01T00:00:00Z',
    });
    const result2 = deriveJugglingOrbitFormalization({
      ...CASCADE_3,
      generatedAt: '2026-01-01T00:00:00Z',
    });
    expect(result1.normalizedThrows).toEqual(result2.normalizedThrows);
    expect(result1.transitions).toEqual(result2.transitions);
    expect(result1.orbits).toEqual(result2.orbits);
    expect(result1.aggregate).toEqual(result2.aggregate);
  });
});
