// @a0n/.ignored_gnosis/src/juggling-orbit-formalization.ts
import {
  type JugglingHand,
  positiveModulo,
  normalizePositiveInt,
  normalizeNonNegativeInt,
  normalizeFinite,
  normalizeHand,
  oppositeHand,
  quantize,
  normalizeSpeed,
  normalizeLabel,
  nowIso,
} from './juggling-utils';

export type { JugglingHand } from './juggling-utils';

export const JUGGLING_ORBIT_THEOREM_IDS = [
  'THM-JUGGLING-ORBIT-TRANSITION-DETERMINISTIC',
  'THM-JUGGLING-ORBIT-DECOMPOSITION-FOLD',
  'THM-JUGGLING-ASYMMETRIC-MULTIPLEX-WITNESS',
  'THM-JUGGLING-CONSERVATION-PER-BEAT-HAND',
  'THM-JUGGLING-MACHINE-ONLY-CANDIDATE-BOUNDARY',
] as const;

export type JugglingOrbitTheoremId =
  (typeof JUGGLING_ORBIT_THEOREM_IDS)[number];

export interface JugglingOrbitThrowSignal {
  readonly beat: number;
  readonly fromHand: JugglingHand;
  readonly flightBeats: number;
  readonly toHand?: JugglingHand;
  readonly lane?: number;
  readonly angleDeg?: number;
  readonly speed?: number;
  readonly label?: string;
}

export interface JugglingOrbitFormalizationInput {
  readonly period: number;
  readonly throws: readonly JugglingOrbitThrowSignal[];
  readonly generatedAt?: string;
  readonly source?: string;
  readonly machineOnlyMultiplexThreshold?: number;
  readonly machineOnlyAngleThresholdDeg?: number;
  readonly machineOnlySpeedRatioThreshold?: number;
}

export interface JugglingOrbitNormalizedThrow {
  readonly id: string;
  readonly beat: number;
  readonly fromHand: JugglingHand;
  readonly toHand: JugglingHand;
  readonly flightBeats: number;
  readonly lane: number;
  readonly angleDeg: number;
  readonly speed: number;
  readonly label: string;
}

export interface JugglingOrbitBeatHandBalance {
  readonly beat: number;
  readonly hand: JugglingHand;
  readonly incoming: number;
  readonly outgoing: number;
  readonly mismatch: number;
  readonly balanced: boolean;
}

export type JugglingOrbitTransitionMapping = 'direct' | 'cycled' | 'open';

export interface JugglingOrbitTransition {
  readonly fromThrowId: string;
  readonly toThrowId: string | null;
  readonly catchBeat: number;
  readonly catchHand: JugglingHand;
  readonly mapping: JugglingOrbitTransitionMapping;
}

export interface JugglingAsymmetricMultiplexWitness {
  readonly witnessId: string;
  readonly beat: number;
  readonly hand: JugglingHand;
  readonly outgoingCount: number;
  readonly throwIds: readonly string[];
  readonly angleRangeDeg: number;
  readonly speedRatio: number;
  readonly asymmetric: boolean;
  readonly machineOnlyCandidate: boolean;
}

export interface JugglingOrbitClass {
  readonly orbitId: string;
  readonly throwIds: readonly string[];
  readonly closed: boolean;
  readonly cycleStartIndex: number;
  readonly cycleLength: number;
  readonly handSignature: string;
  readonly beatSignature: string;
  readonly asymmetricMultiplex: boolean;
  readonly machineOnlyCandidate: boolean;
}

export interface JugglingOrbitFormalizationAggregate {
  readonly throwCount: number;
  readonly transitionCount: number;
  readonly openTransitionCount: number;
  readonly orbitCount: number;
  readonly closedOrbitCount: number;
  readonly openOrbitCount: number;
  readonly asymmetricMultiplexOrbitCount: number;
  readonly machineOnlyOrbitCount: number;
  readonly maxMultiplexPerBeatHand: number;
  readonly conservationMismatchCount: number;
}

export interface JugglingOrbitFormalizationBundle {
  readonly generatedAt: string;
  readonly source: string;
  readonly theoremIds: readonly JugglingOrbitTheoremId[];
  readonly period: number;
  readonly normalizedThrows: readonly JugglingOrbitNormalizedThrow[];
  readonly beatHandBalances: readonly JugglingOrbitBeatHandBalance[];
  readonly asymmetricMultiplexWitnesses: readonly JugglingAsymmetricMultiplexWitness[];
  readonly transitions: readonly JugglingOrbitTransition[];
  readonly orbits: readonly JugglingOrbitClass[];
  readonly aggregate: JugglingOrbitFormalizationAggregate;
}

interface CanonicalThrow {
  readonly beat: number;
  readonly fromHand: JugglingHand;
  readonly toHand: JugglingHand;
  readonly flightBeats: number;
  readonly lane: number;
  readonly angleDeg: number;
  readonly speed: number;
  readonly label: string;
}

interface MultiplexThresholds {
  readonly multiplex: number;
  readonly angleDeg: number;
  readonly speedRatio: number;
}

const HAND_ORDER: Record<JugglingHand, number> = {
  left: 0,
  right: 1,
};

// Utility functions imported from juggling-utils.ts

function compareNumbers(left: number, right: number): number {
  return left - right;
}

function compareHands(left: JugglingHand, right: JugglingHand): number {
  return HAND_ORDER[left] - HAND_ORDER[right];
}

function throwSignature(throwSignal: CanonicalThrow): string {
  return [
    `b${throwSignal.beat}`,
    `f${throwSignal.fromHand}`,
    `t${throwSignal.toHand}`,
    `l${throwSignal.lane}`,
    `h${throwSignal.flightBeats}`,
    `a${throwSignal.angleDeg.toFixed(3)}`,
    `s${throwSignal.speed.toFixed(3)}`,
    throwSignal.label,
  ].join(':');
}

function compareCanonicalThrows(
  left: CanonicalThrow,
  right: CanonicalThrow
): number {
  return (
    compareNumbers(left.beat, right.beat) ||
    compareHands(left.fromHand, right.fromHand) ||
    compareNumbers(left.lane, right.lane) ||
    compareNumbers(left.flightBeats, right.flightBeats) ||
    compareHands(left.toHand, right.toHand) ||
    compareNumbers(left.angleDeg, right.angleDeg) ||
    compareNumbers(left.speed, right.speed) ||
    left.label.localeCompare(right.label)
  );
}

function compareIncomingThrows(
  left: JugglingOrbitNormalizedThrow,
  right: JugglingOrbitNormalizedThrow
): number {
  return (
    compareNumbers(left.beat, right.beat) ||
    compareHands(left.fromHand, right.fromHand) ||
    compareNumbers(left.lane, right.lane) ||
    compareNumbers(left.flightBeats, right.flightBeats) ||
    compareHands(left.toHand, right.toHand) ||
    compareNumbers(left.angleDeg, right.angleDeg) ||
    compareNumbers(left.speed, right.speed) ||
    left.id.localeCompare(right.id)
  );
}

function compareOutgoingThrows(
  left: JugglingOrbitNormalizedThrow,
  right: JugglingOrbitNormalizedThrow
): number {
  return (
    compareNumbers(left.lane, right.lane) ||
    compareNumbers(left.flightBeats, right.flightBeats) ||
    compareHands(left.toHand, right.toHand) ||
    compareNumbers(left.angleDeg, right.angleDeg) ||
    compareNumbers(left.speed, right.speed) ||
    left.id.localeCompare(right.id)
  );
}

function beatHandKey(beat: number, hand: JugglingHand): string {
  return `${beat}|${hand}`;
}

function parseBeatHandKey(key: string): { beat: number; hand: JugglingHand } {
  const [beatText, handText] = key.split('|');
  const beat = Number.parseInt(beatText ?? '0', 10);
  const hand = normalizeHand(handText as JugglingHand);
  return {
    beat: Number.isFinite(beat) ? beat : 0,
    hand,
  };
}

function compareBeatHandKeys(left: string, right: string): number {
  const leftParsed = parseBeatHandKey(left);
  const rightParsed = parseBeatHandKey(right);
  return (
    compareNumbers(leftParsed.beat, rightParsed.beat) ||
    compareHands(leftParsed.hand, rightParsed.hand)
  );
}

function normalizeThrows(
  throwsSignals: readonly JugglingOrbitThrowSignal[],
  period: number
): readonly JugglingOrbitNormalizedThrow[] {
  const canonical = throwsSignals
    .map((signal, index) => {
      const beat = positiveModulo(Math.floor(signal.beat), period);
      const fromHand = normalizeHand(signal.fromHand);
      const flightBeats = normalizePositiveInt(signal.flightBeats, 1);
      const lane = normalizeNonNegativeInt(signal.lane ?? 0, 0);
      const angleDeg = quantize(normalizeFinite(signal.angleDeg, 0), 3);
      const speed = normalizeSpeed(signal.speed);
      const toHand = signal.toHand
        ? normalizeHand(signal.toHand)
        : flightBeats % 2 === 0
          ? fromHand
          : oppositeHand(fromHand);
      const label = normalizeLabel(signal.label, `throw-${index + 1}`);

      return {
        beat,
        fromHand,
        toHand,
        flightBeats,
        lane,
        angleDeg,
        speed,
        label,
      } satisfies CanonicalThrow;
    })
    .sort(compareCanonicalThrows);

  const occurrences = new Map<string, number>();
  return canonical.map((entry) => {
    const signature = throwSignature(entry);
    const count = (occurrences.get(signature) ?? 0) + 1;
    occurrences.set(signature, count);
    return {
      id: `throw:${signature}:n${count}`,
      ...entry,
    };
  });
}

function buildBeatHandBalance(
  incomingByCatchKey: ReadonlyMap<string, readonly JugglingOrbitNormalizedThrow[]>,
  outgoingByThrowKey: ReadonlyMap<string, readonly JugglingOrbitNormalizedThrow[]>
): readonly JugglingOrbitBeatHandBalance[] {
  const keys = new Set<string>([
    ...incomingByCatchKey.keys(),
    ...outgoingByThrowKey.keys(),
  ]);

  return [...keys].sort(compareBeatHandKeys).map((key) => {
    const parsed = parseBeatHandKey(key);
    const incoming = incomingByCatchKey.get(key)?.length ?? 0;
    const outgoing = outgoingByThrowKey.get(key)?.length ?? 0;
    const mismatch = outgoing - incoming;

    return {
      beat: parsed.beat,
      hand: parsed.hand,
      incoming,
      outgoing,
      mismatch,
      balanced: mismatch === 0,
    };
  });
}

function resolveThresholds(
  input: JugglingOrbitFormalizationInput
): MultiplexThresholds {
  return {
    multiplex: normalizePositiveInt(
      input.machineOnlyMultiplexThreshold ?? 2,
      2
    ),
    angleDeg: Math.max(0, normalizeFinite(input.machineOnlyAngleThresholdDeg, 12)),
    speedRatio: Math.max(
      1,
      normalizeFinite(input.machineOnlySpeedRatioThreshold, 1.3)
    ),
  };
}

function buildMultiplexWitnesses(
  outgoingByThrowKey: ReadonlyMap<string, readonly JugglingOrbitNormalizedThrow[]>,
  thresholds: MultiplexThresholds
): readonly JugglingAsymmetricMultiplexWitness[] {
  const keys = [...outgoingByThrowKey.keys()].sort(compareBeatHandKeys);

  const witnesses: JugglingAsymmetricMultiplexWitness[] = [];
  for (const key of keys) {
    const group = [...(outgoingByThrowKey.get(key) ?? [])].sort(
      compareOutgoingThrows
    );
    if (group.length < 2) {
      continue;
    }

    const parsed = parseBeatHandKey(key);
    const angleValues = group.map((entry) => entry.angleDeg);
    const speedValues = group.map((entry) => entry.speed);
    const minAngle = Math.min(...angleValues);
    const maxAngle = Math.max(...angleValues);
    const minSpeed = Math.min(...speedValues);
    const maxSpeed = Math.max(...speedValues);
    const angleRangeDeg = quantize(maxAngle - minAngle, 3);
    const speedRatio =
      minSpeed > 0 ? quantize(maxSpeed / minSpeed, 3) : Number.POSITIVE_INFINITY;
    const asymmetric = angleRangeDeg > 0 || speedRatio > 1;
    const machineOnlyCandidate =
      asymmetric &&
      group.length >= thresholds.multiplex &&
      (angleRangeDeg >= thresholds.angleDeg || speedRatio >= thresholds.speedRatio);

    witnesses.push({
      witnessId: `witness:b${parsed.beat}:h${parsed.hand}:m${group.length}`,
      beat: parsed.beat,
      hand: parsed.hand,
      outgoingCount: group.length,
      throwIds: group.map((entry) => entry.id),
      angleRangeDeg,
      speedRatio,
      asymmetric,
      machineOnlyCandidate,
    });
  }

  return witnesses;
}

function deriveOrbitClasses(
  throwsSignals: readonly JugglingOrbitNormalizedThrow[],
  transitionByThrowId: ReadonlyMap<string, string | null>,
  asymmetricThrowIds: ReadonlySet<string>,
  machineOnlyThrowIds: ReadonlySet<string>
): readonly JugglingOrbitClass[] {
  const classes: JugglingOrbitClass[] = [];
  const visited = new Set<string>();
  const throwById = new Map(throwsSignals.map((entry) => [entry.id, entry]));

  for (const entry of throwsSignals) {
    if (visited.has(entry.id)) {
      continue;
    }

    const localIndex = new Map<string, number>();
    const path: string[] = [];
    let current: string | null = entry.id;

    while (current !== null && !localIndex.has(current) && !visited.has(current)) {
      localIndex.set(current, path.length);
      path.push(current);
      current = transitionByThrowId.get(current) ?? null;
    }

    for (const throwId of path) {
      visited.add(throwId);
    }

    const closed = current !== null && localIndex.has(current);
    const cycleStartIndex =
      closed && current !== null ? (localIndex.get(current) ?? 0) : 0;
    const cycleLength = closed ? path.length - cycleStartIndex : 0;
    const classThrows = path
      .map((throwId) => throwById.get(throwId))
      .filter((throwSignal): throwSignal is JugglingOrbitNormalizedThrow =>
        Boolean(throwSignal)
      );

    const handSignature = classThrows
      .map((throwSignal) => (throwSignal.fromHand === 'left' ? 'L' : 'R'))
      .join('');
    const beatSignature = classThrows
      .map((throwSignal) => String(throwSignal.beat))
      .join('-');
    const asymmetricMultiplex = path.some((throwId) =>
      asymmetricThrowIds.has(throwId)
    );
    const machineOnlyCandidate = path.some((throwId) =>
      machineOnlyThrowIds.has(throwId)
    );

    classes.push({
      orbitId: `orbit:${classes.length + 1}`,
      throwIds: path,
      closed,
      cycleStartIndex,
      cycleLength,
      handSignature,
      beatSignature,
      asymmetricMultiplex,
      machineOnlyCandidate,
    });
  }

  return classes;
}

/**
 * Handles the gnosis derive Juggling Orbit Formalization workflow.
 */
export function deriveJugglingOrbitFormalization(
  input: JugglingOrbitFormalizationInput
): JugglingOrbitFormalizationBundle {
  const period = normalizePositiveInt(input.period, 1);
  const normalizedThrows = normalizeThrows(input.throws, period);
  const thresholds = resolveThresholds(input);

  const incomingByCatchKey = new Map<string, JugglingOrbitNormalizedThrow[]>();
  const outgoingByThrowKey = new Map<string, JugglingOrbitNormalizedThrow[]>();
  const outgoingCountByBeatHand = new Map<string, number>();

  for (const throwSignal of normalizedThrows) {
    const outgoingKey = beatHandKey(throwSignal.beat, throwSignal.fromHand);
    const catchBeat = positiveModulo(throwSignal.beat + throwSignal.flightBeats, period);
    const incomingKey = beatHandKey(catchBeat, throwSignal.toHand);

    const outgoing = outgoingByThrowKey.get(outgoingKey);
    if (outgoing) {
      outgoing.push(throwSignal);
    } else {
      outgoingByThrowKey.set(outgoingKey, [throwSignal]);
    }

    const incoming = incomingByCatchKey.get(incomingKey);
    if (incoming) {
      incoming.push(throwSignal);
    } else {
      incomingByCatchKey.set(incomingKey, [throwSignal]);
    }
  }

  for (const [key, group] of outgoingByThrowKey.entries()) {
    outgoingCountByBeatHand.set(key, group.length);
  }

  const beatHandBalances = buildBeatHandBalance(incomingByCatchKey, outgoingByThrowKey);
  const multiplexWitnesses = buildMultiplexWitnesses(outgoingByThrowKey, thresholds);
  const asymmetricThrowIds = new Set<string>();
  const machineOnlyThrowIds = new Set<string>();

  for (const witness of multiplexWitnesses) {
    if (witness.asymmetric) {
      for (const throwId of witness.throwIds) {
        asymmetricThrowIds.add(throwId);
      }
    }
    if (witness.machineOnlyCandidate) {
      for (const throwId of witness.throwIds) {
        machineOnlyThrowIds.add(throwId);
      }
    }
  }

  const transitions: JugglingOrbitTransition[] = [];
  const transitionByThrowId = new Map<string, string | null>();
  const keys = [...incomingByCatchKey.keys()].sort(compareBeatHandKeys);

  for (const key of keys) {
    const parsed = parseBeatHandKey(key);
    const incoming = [...(incomingByCatchKey.get(key) ?? [])].sort(
      compareIncomingThrows
    );
    const outgoing = [...(outgoingByThrowKey.get(key) ?? [])].sort(
      compareOutgoingThrows
    );

    for (let index = 0; index < incoming.length; index += 1) {
      const source = incoming[index];
      if (!source) {
        continue;
      }

      if (outgoing.length === 0) {
        transitions.push({
          fromThrowId: source.id,
          toThrowId: null,
          catchBeat: parsed.beat,
          catchHand: parsed.hand,
          mapping: 'open',
        });
        transitionByThrowId.set(source.id, null);
        continue;
      }

      const targetIndex = index % outgoing.length;
      const target = outgoing[targetIndex];
      if (!target) {
        transitions.push({
          fromThrowId: source.id,
          toThrowId: null,
          catchBeat: parsed.beat,
          catchHand: parsed.hand,
          mapping: 'open',
        });
        transitionByThrowId.set(source.id, null);
        continue;
      }

      transitions.push({
        fromThrowId: source.id,
        toThrowId: target.id,
        catchBeat: parsed.beat,
        catchHand: parsed.hand,
        mapping: index < outgoing.length ? 'direct' : 'cycled',
      });
      transitionByThrowId.set(source.id, target.id);
    }
  }

  for (const throwSignal of normalizedThrows) {
    if (!transitionByThrowId.has(throwSignal.id)) {
      transitions.push({
        fromThrowId: throwSignal.id,
        toThrowId: null,
        catchBeat: positiveModulo(throwSignal.beat + throwSignal.flightBeats, period),
        catchHand: throwSignal.toHand,
        mapping: 'open',
      });
      transitionByThrowId.set(throwSignal.id, null);
    }
  }

  const sortedTransitions = [...transitions].sort((left, right) => {
    const leftThrow = normalizedThrows.find((entry) => entry.id === left.fromThrowId);
    const rightThrow = normalizedThrows.find((entry) => entry.id === right.fromThrowId);
    return (
      compareNumbers(leftThrow?.beat ?? 0, rightThrow?.beat ?? 0) ||
      compareHands(leftThrow?.fromHand ?? 'left', rightThrow?.fromHand ?? 'left') ||
      left.fromThrowId.localeCompare(right.fromThrowId)
    );
  });

  const orbits = deriveOrbitClasses(
    normalizedThrows,
    transitionByThrowId,
    asymmetricThrowIds,
    machineOnlyThrowIds
  );

  const aggregate: JugglingOrbitFormalizationAggregate = {
    throwCount: normalizedThrows.length,
    transitionCount: sortedTransitions.length,
    openTransitionCount: sortedTransitions.filter(
      (transition) => transition.toThrowId === null
    ).length,
    orbitCount: orbits.length,
    closedOrbitCount: orbits.filter((orbit) => orbit.closed).length,
    openOrbitCount: orbits.filter((orbit) => !orbit.closed).length,
    asymmetricMultiplexOrbitCount: orbits.filter(
      (orbit) => orbit.asymmetricMultiplex
    ).length,
    machineOnlyOrbitCount: orbits.filter((orbit) => orbit.machineOnlyCandidate)
      .length,
    maxMultiplexPerBeatHand: [...outgoingCountByBeatHand.values()].reduce(
      (max, value) => Math.max(max, value),
      0
    ),
    conservationMismatchCount: beatHandBalances.filter(
      (entry) => !entry.balanced
    ).length,
  };

  return {
    generatedAt: input.generatedAt ?? nowIso(),
    source: input.source ?? 'jugglingOrbitFormalization/v1',
    theoremIds: JUGGLING_ORBIT_THEOREM_IDS,
    period,
    normalizedThrows,
    beatHandBalances,
    asymmetricMultiplexWitnesses: multiplexWitnesses,
    transitions: sortedTransitions,
    orbits,
    aggregate,
  };
}
