// @a0n/.ignored_gnosis/src/knapsack-viewport-topology.ts
export interface ViewportPackingResult {
  pleromicData: number;
  screenCapacity: number;
  visualOverlap: number;
  isCluttered: boolean;
  isValidProjection: boolean;
}

/**
 * THM-VIEWPORT-PACKING-PROJECTION: The screen view is a rigid 2D Base Space.
 * The Knapsack algorithm optimizes the Cut-and-Project angle of higher-dimensional
 * Pleromic data to maximize density without triggering visual Reidemeister Slams
 * (clutter/overlap).
 */
export function evaluateViewportPacking(
  pleromicData: number,
  screenCapacity: number,
  visualOverlap?: number
): ViewportPackingResult {
  const isCluttered = pleromicData > screenCapacity;
  const computedVisualOverlap = isCluttered ? pleromicData - screenCapacity : 0;

  if (visualOverlap !== undefined && computedVisualOverlap !== visualOverlap) {
    throw new Error(
      `Viewport topology boundary broken: visualOverlap provided (${visualOverlap}) does not match required delta (${computedVisualOverlap})`
    );
  }

  const finalVisualOverlap = visualOverlap ?? computedVisualOverlap;

  return {
    pleromicData,
    screenCapacity,
    visualOverlap: finalVisualOverlap,
    isCluttered,
    isValidProjection: finalVisualOverlap === 0 && pleromicData <= screenCapacity,
  };
}
