// @a0n/.ignored_gnosis/src/knotgraft-syzygy-cannon.ts
/**
 * Knotgraft Syzygy Cannon — thin knotgraft-flavored wrapper around
 * `@a0n/aeon/fabric/client`'s `AeonFabricCannon`.
 *
 * The cannon pattern (preconnect, arm, barrier-launch, race, dedup at
 * receiver) now lives in aeon. This module contributes only the
 * knotgraft-specific pieces:
 *
 *   - Bearer-token injection via the http lane
 *   - `POST /api/knotgraft?k=<n>[&emitSql=1]` path construction
 *   - Lacey-cipher derivation from a passphrase
 *   - Mapping raw payload bytes → Web `Request` for the cannon
 *   - Projecting the cannon's generic `Response` + leg report back into
 *     the `KnotgraftResponse` / knotgraft-shaped report consumers expect
 */

import {
  AeonFabricCannon,
  type CannonConfig,
  type CannonLegReport,
  type FabricTransport,
} from '@a0n/aeon/fabric/client';
import {
  rotatingHelixEncryptFrame,
  rotatingHelixDecryptFrame,
  rotatingHelixKeyFromSeed,
  seedFromPassphrase,
} from './bitwise-cipher';
import type { KnotgraftResponse } from './knotgraft-client';

// ─── Config (back-compat surface) ───────────────────────────────────

export type CannonTransport = FabricTransport;

export interface SyzygyCannonConfig {
  readonly http: { readonly baseUrl: string; readonly token: string };
  readonly httpRust?: { readonly baseUrl: string; readonly token?: string };
  readonly tcpFlow?: { readonly host: string; readonly port: number };
  readonly udpFlow?: { readonly host: string; readonly port: number };
  readonly udpFlowRust?: { readonly host: string; readonly port: number };
  readonly legTimeoutMs?: number;
  readonly laceyCipher?: {
    readonly enabled?: boolean;
    readonly passphrase?: string;
  };
}

export { type CannonLegReport };

export interface CannonFireResult {
  readonly result: KnotgraftResponse;
  readonly winner: CannonTransport;
  readonly winnerElapsedMs: number;
  readonly legs: ReadonlyArray<CannonLegReport>;
}

export interface FireOptions {
  readonly laneCount?: number;
  readonly emitSql?: boolean;
}

// ─── Cannon ─────────────────────────────────────────────────────────

export class KnotgraftSyzygyCannon {
  private readonly config: SyzygyCannonConfig;
  private readonly baseUrl: string;
  private readonly bearer: string;
  private inner: AeonFabricCannon | null = null;

  constructor(config: SyzygyCannonConfig) {
    if (!config.http?.token) throw new Error('cannon requires http.token');
    this.config = config;
    this.baseUrl = config.http.baseUrl;
    this.bearer = config.http.token;
  }

  async arm(): Promise<void> {
    // Resolve Lacey cipher from passphrase if enabled, then build
    // the inner cannon with the cipher baked in.
    const enabled = this.config.laceyCipher?.enabled ?? true;
    const passphrase =
      this.config.laceyCipher?.passphrase ??
      process.env['KNOTGRAFT_CIPHER_PASSPHRASE'] ??
      null;
    const cipher =
      enabled && passphrase
        ? await (async () => {
            const seed = await seedFromPassphrase(passphrase);
            const key = rotatingHelixKeyFromSeed(seed);
            return {
              encryptFrame: (frame: Uint8Array) => rotatingHelixEncryptFrame(frame, key),
              decryptFrame: (frame: Uint8Array) => rotatingHelixDecryptFrame(frame, key),
            };
          })()
        : undefined;

    const innerConfig: CannonConfig = {
      http: { baseUrl: this.config.http.baseUrl, bearer: this.config.http.token },
      httpAlt: this.config.httpRust
        ? {
            baseUrl: this.config.httpRust.baseUrl,
            bearer: this.config.httpRust.token ?? this.config.http.token,
          }
        : undefined,
      tcpFlow: this.config.tcpFlow,
      udpFlow: this.config.udpFlow,
      udpFlowAlt: this.config.udpFlowRust,
      cipher,
      legTimeoutMs: this.config.legTimeoutMs ?? 10_000,
    };
    this.inner = new AeonFabricCannon(innerConfig);
    await this.inner.arm();
  }

  /** Fire raw ingest bytes simultaneously over every armed transport. */
  async fire(bytes: Uint8Array, options: FireOptions = {}): Promise<CannonFireResult> {
    if (!this.inner) {
      throw new Error('cannon not armed — call arm() before fire()');
    }
    const laneCount = options.laneCount ?? 16;
    const emitSql = options.emitSql ?? false;
    const path = `/api/knotgraft?k=${laneCount}${emitSql ? '&emitSql=1' : ''}`;
    const request = new Request(new URL(path, this.baseUrl), {
      method: 'POST',
      headers: {
        'content-type': 'application/octet-stream',
        authorization: `Bearer ${this.bearer}`,
      },
      body: bytes as BodyInit,
    });

    const fabricResult = await this.inner.fire(request);
    const result = (await fabricResult.response.json()) as KnotgraftResponse;
    return {
      result,
      winner: fabricResult.winner,
      winnerElapsedMs: fabricResult.winnerElapsedMs,
      legs: fabricResult.legs,
    };
  }

  async close(): Promise<void> {
    if (this.inner) {
      await this.inner.close();
      this.inner = null;
    }
  }
}
