// @a0n/.ignored_gnosis/src/lacey-nested-prng.ts
/**
 * Lacey-style nested PRNG — mirrors rotating half-helix geometry
 * (`rotatingHelixEncryptFrame` in `bitwise-cipher.ts`).
 *
 * Lean: `LaceyNestedRotation` (geometry), `AperiodicPseudorandom` (Weyl ℝ story),
 * `CryptographicPseudorandom` (CSPRNG *definition*: negligible advantage). This
 * file is not a CSPRNG.
 *
 * Layers (inside → out):
 * 1. **Dual half-plane** — α vs β XOR seeds (same role as payload XOR base).
 * 2. **Per-dimension phase** — 54 `phaseOffsets[i]` expanded from the root seed.
 * 3. **Helical advance** — `rotation * 7920` matches the cipher’s phase rotation.
 * 4. **Outer rotation index** — `rotationAt(position)` with `HELIX_TURN_SAMPLES`.
 * 5. **Weyl φ⁻¹ injection** — low-discrepancy additive term (FORMAL_LEDGER Pass 68–70).
 * 6. **SplitMix64 avalanche** — final byte mixing (same family as Murmur/SplitMix).
 *
 * Simulation/procedural only; same non-crypto stance as `Xorshift64` in `bitwise-cipher.ts`.
 *
 * @module @a0n/gnosis/lacey-nested-prng
 */

import {
  HELIX_TURN_SAMPLES,
  Xorshift64,
  isAlphaHalf,
  rotationAt,
} from './bitwise-cipher';

const MASK = 0xFFFFFFFFFFFFFFFFn;
const HELIX_ADVANCE = 7920n;

/** φ⁻¹ · 2^64 rounded — pairs with `AperiodicPseudorandom` real parameter story. */
const INV_PHI_U64 = 11400714819323198485n;

const M0 = 0x9e3779b97f4a7c15n;
const M1 = 0x517cc1b727220a95n;

function splitMix64(z: bigint): bigint {
  let x = z & MASK;
  x = (x ^ (x >> 30n)) * 0xbf58476d1ce4e5b9n;
  x &= MASK;
  x = (x ^ (x >> 27n)) * 0x94d049bb133111ebn;
  x &= MASK;
  x ^= x >> 31n;
  return x & MASK;
}

/**
 * Nested Lacey / Weyl / SplitMix PRNG — mirrors cipher geometry without encrypting frames.
 */
export class LaceyNestedRng {
  private s0: bigint;
  private s1: bigint;
  private counter: bigint;
  private readonly phaseOffsets: Uint32Array;
  private readonly xorSeedAlpha: bigint;
  private readonly xorSeedBeta: bigint;

  constructor(seed: bigint) {
    const gen = new Xorshift64(seed === 0n ? 0xdeadbeefcafebaben : seed);
    const phases = new Uint32Array(54);
    for (let i = 0; i < 54; i++) {
      phases[i] = Number(gen.next() & 0xffffffffn);
    }
    this.phaseOffsets = phases;
    this.xorSeedAlpha = gen.next();
    this.xorSeedBeta = gen.next();
    this.s0 = gen.next();
    this.s1 = gen.next();
    this.counter = 0n;
  }

  /** Monotonic step index (FlowFrame `position` analogue). */
  get stepIndex(): bigint {
    return this.counter;
  }

  private advanceCounter(): number {
    this.counter += 1n;
    return Number(this.counter & 0xffff_ffffn);
  }

  /** 64-bit word — nested rotation + keyed mix + avalanche. */
  nextU64(): bigint {
    const pos = this.advanceCounter();
    const rot = rotationAt(pos);
    const dim = Number((this.s0 ^ this.s1) & 63n) % 54;
    const beta = !isAlphaHalf(dim, rot);
    const phaseOffset = this.phaseOffsets[dim];
    if (phaseOffset === undefined) {
      throw new Error(`Missing phase offset for dimension ${dim}`);
    }
    const phase = BigInt(phaseOffset >>> 0);
    const weyl = (INV_PHI_U64 * BigInt(pos)) & MASK;

    let mix = this.s0 ^ (this.s1 << 13n);
    mix = (mix + HELIX_ADVANCE * BigInt(rot) + phase + weyl) & MASK;

    const base = beta ? this.xorSeedBeta : this.xorSeedAlpha;
    const keyed =
      (base ^ BigInt(dim) * M0 ^ BigInt(rot) * M1) & MASK;

    const out = splitMix64(mix ^ keyed);
    this.s0 = this.s1;
    this.s1 = out;
    return out;
  }

  /** [0, 1) float, upper exclusive. */
  nextFloat(): number {
    return Number(this.nextU64() >> 32n) / 2 ** 32;
  }
}

/**
 * Factory — `seed` is any non-zero bigint; zero is mapped to a fixed nonzero constant.
 */
export function createLaceyNestedRng(seed: bigint): LaceyNestedRng {
  return new LaceyNestedRng(seed);
}

/** Re-export helix constants for callers composing schedules with the PRNG. */
export { HELIX_TURN_SAMPLES, rotationAt, isAlphaHalf };
