// @a0n/.ignored_gnosis/src/lib.ts
// gnosis-math — the mathematical kernel
export {
  PRIMITIVES,
  type Primitive,
  GNOSTIC,
  PHI,
  INV_PHI,
  SQRT5,
  fib,
  triangular,
  pairwise,
  BOSONS,
  type Boson,
  type BosonFamily,
  buleyeanWeight,
  complementDistributionWeights,
  pipelineEnergy,
  isConfined,
  verifyTripleCoincidence,
  verifyCassini,
  // Time system
  PICOLORENZO_SECONDS,
  PICOLORENZO_MS,
  NANOLORENZO_SECONDS,
  AEON_SECONDS,
  type GnosticTimestamp,
  nowPicolorenzos,
  toPicolorenzos,
  fromPicolorenzos,
  epochMsToPicolorenzos,
  picolorenzosToEpochMs,
  epochSecondsToPicolorenzos,
  picolorenzosToEpochSeconds,
  parseToGnostic,
  dateToGnostic,
  epochToGnostic,
  epochMsToGnostic,
  gnosticToIso,
  gnosticToEpoch,
  formatPicolorenzos,
  decomposeGnosticTime,
  formatGnosticTime,
  formatDual,
  secondsToPicolorenzos,
  picolorenzosToSeconds,
  durationPicolorenzos,
  durationFromIso,
  // Solar geometry (core Gnostic Time System)
  OBLIQUITY_DEG,
  type TwilightZone,
  type SolarObserver,
  type SolarPosition,
  type SolarPhase,
  type SolarHalfDay,
  solarDeclination,
  equationOfTime,
  classifyTwilightZone,
  twilightZoneFraction,
  solarPosition,
  isDaylightAtPLo,
  solarPositionAtPLo,
  solarPhase,
  solarPhaseContinuous,
  solarHalfDay,
  MINUTES_PER_PICOLORENZO,
  // Physics constants
  GRAVITATIONAL_CONSTANT_SI,
  SOLAR_MASS_KG,
  SPEED_OF_LIGHT_MS,
  LIGHT_YEAR_METERS,
  SCHWARZSCHILD_RADIUS_SOLAR_LY,
  HUBBLE_CONSTANT_KM_S_MPC,
  MPC_IN_LY,
  // Light-picolorenzo (causal unit)
  C_LY_PER_SECOND,
  C_AU_PER_SECOND,
  LIGHT_PICOLORENZO_LY,
  LIGHT_PICOLORENZO_AU,
  lightYearsToLightPicolorenzos,
  lightPicolorenzosToLightYears,
  lightTravelTimePLo,
  isCausallyReachable,
  CIRCADIAN_DRIFT_RATIO,
  AEON_KENOMA_IDENTITY,
  RESPIRATORY_RESONANCE,
  RESPIRATORY_RESONANCE_BY_AGE,
} from './math.js';

export {
  JULIAN_YEAR_SECONDS,
  julianYearsToPicolorenzos,
  EARTH_FORMATION_AGE_GYR,
  EARTH_FORMATION_AGE_YEARS,
  EARTH_FORMATION_AGE_YEARS_UNCERTAINTY_ONE_SIGMA,
  NANOLORENZOS_PER_PICOLORENZO,
  type EarthFormationGnosticReport,
  earthFormationAgeGnosticReport,
  earthFormationAgeUncertaintyBracketGnostic,
} from './earth-chronometry.js';

export { CrossPollinationEngine } from './CrossPollinationEngine.js';
export { StressEnergyMonitor } from './StressEnergyMonitor.js';
export * from './universe/index.js';
export * from './iota-trees.js';
export * from './protocol69.js';
export * from './protocol69-r1-witness.js';
export * from './protocol69-r1-phyle-entropy.js';
export * from './protocol69-r1-worksteal.js';
export * from './sentience.js';
export * from './probability-certificate.js';
export * from './probability-routing.js';
export {
  initGnosisTracingFromEnv,
  isGnosisTracingEnabled,
  withGnosisSpan,
  tracer as gnosisTracer,
} from './apm.js';
export {
  appendCertifiedAppEnvelopeAttestation,
  appendCertifiedAppEnvelopeAttestationToFile,
  appendCertifiedAppEnvelopeAttestationToQDoc,
  auditCertifiedAppEnvelope,
  createCertifiedAppEnvelope,
  createCertifiedAppEnvelopeFromGgSource,
  createCertifiedAppEnvelopeMirror,
  loadCertifiedAppEnvelopeByAliasFromQDoc,
  loadCertifiedAppEnvelopeFromFile,
  loadCertifiedAppEnvelopeFromQDoc,
  saveCertifiedAppEnvelopeToFile,
  serializeCertifiedAppEnvelopeMirror,
  storeCertifiedAppEnvelopeInQDoc,
  type CertifiedAppAeonPidRegistrationAttestation,
  type CertifiedAppAgentSwarmAdjudicationAttestation,
  type CertifiedAppAgentSwarmAdjudicationSummary,
  type CertifiedAppAttestationKind,
  type CertifiedAppDeployBuildAttestation,
  type CertifiedAppEnvelope,
  type CertifiedAppEnvelopeAttestation,
  type CertifiedAppEnvelopeAudit,
  type CertifiedAppEnvelopeCore,
  type CertifiedAppEnvelopeMirror,
  type CertifiedAppJusticeGateAttestation,
  type CertifiedAppPublishProofAttestation,
  type CertifiedAppRuntimeStartAttestation,
  type CertifiedAppVfsSnapshotAttestation,
  type CreateCertifiedAppEnvelopeFromGgSourceInput,
  type CreateCertifiedAppEnvelopeInput,
} from './certified-app.js';
export {
  DARK_SHROUD_THEOREM_IDS,
  auditDarkShroudBundle,
  deriveDarkShroud,
  type DarkShroudBundle,
  type DarkShroudBundleAudit,
  type DarkShroudDecision,
  type DarkShroudEvidenceMode,
  type DarkShroudRouteInput,
  type DarkShroudShroudMode,
  type DarkShroudTheoremId,
  type DeriveDarkShroudInput,
} from './dark-shroud.js';
export {
  XRANGER_BOUNDARY_ACTIONS,
  XRANGER_BOUNDARY_THEOREM_IDS,
  XRANGER_INSPECTION_ACTIONS,
  auditXRangerBoundaryManifest,
  createXRangerBoundaryManifest,
  floorProbability as xRangerFloorProbability,
  weakestInspectionFloor as xRangerWeakestInspectionFloor,
  type CreateXRangerBoundaryManifestInput,
  type XRangerBoundaryAction,
  type XRangerBoundaryManifest,
  type XRangerBoundaryManifestAudit,
  type XRangerBoundaryManagerContract,
  type XRangerBoundaryShellContract,
  type XRangerBoundaryTheoremId,
  type XRangerDirection,
  type XRangerInspectionAction,
  type XRangerInspectionFloor,
  type XRangerMeshPolicy,
  type XRangerPrismShroud,
  type XRangerStateClass,
  type XRangerTransport,
  type XRangerTrustTier,
} from './x-ranger-boundary.js';
export * from './structured-data.js';
export {
  graphAstToLawProgram,
  ggProgramToLawProgram,
  certifyGnosisAst,
  certifyGnosisProgram,
  createUniverseAdmissionReport,
  createLawStateFromCertificate,
  toLawCapabilitySummary,
  formatUniverseAdmissionReasons,
  assertUniverseAdmission,
  type GnosisLawProgramOptions,
  type GnosisUniverseAdmissionOptions,
} from './law.js';

export {
  EYE_TRACKING_FAILURE_DIMENSIONS,
  EYE_TRACKING_ZONES,
  appendEyeTrackingFailureToQDoc,
  appendEyeTrackingObservationToQDoc,
  auditEyeTrackingSession,
  buildEyeTrackingMerkleTree,
  classifyEyeTrackingZone,
  completeEyeTrackingSession,
  createEyeTrackingCorpus,
  createEyeTrackingBuleyeanTrainingJob,
  createEyeTrackingDatasetSlice,
  createEyeTrackingQDoc,
  createEyeTrackingSession,
  createEyeTrackingTrainingManifest,
  createEyeTrackingVentRecords,
  decodeEyeTrackingFailureEvent,
  defaultEyeTrackingBuleyeanTrainingConfig,
  defaultEyeTrackingCalibrationPlan,
  encodeEyeTrackingFailureEvent,
  exportEyeTrackingBuleyeanPersonalizationDelta,
  ingestEyeTrackingSession,
  loadEyeTrackingSessionByAliasFromQDoc,
  loadEyeTrackingSessionFromQDoc,
  recordEyeTrackingCalibrationSample,
  recordEyeTrackingFailure,
  selectEyeTrackingSessions,
  serializeEyeTrackingRejectionJSONL,
  storeEyeTrackingSessionInQDoc,
  summarizeEyeTrackingBuleyeanTrainingCorpus,
  summarizeEyeTrackingFailurePressure,
  toEyeTrackingRejectionRecord,
  appendEyeTrackingObservation,
  appendEyeTrackingBuleyeanRejectionBatch,
} from './eyetracking.js';
export type {
  EyeTrackingAuditReport,
  EyeTrackingBuleyeanAppendOptions,
  EyeTrackingBuleyeanAppendResult,
  EyeTrackingBuleyeanCorpusSummary,
  EyeTrackingBuleyeanCorpusSummaryOptions,
  EyeTrackingBuleyeanCurriculum,
  EyeTrackingBuleyeanFederatedConfig,
  EyeTrackingBuleyeanPersonalizationDelta,
  EyeTrackingBuleyeanPersonalizationDeltaOptions,
  EyeTrackingBuleyeanSplitSummary,
  EyeTrackingBuleyeanTrainingConfig,
  EyeTrackingBuleyeanTrainingConfigOverrides,
  EyeTrackingBuleyeanTrainingJob,
  EyeTrackingBuleyeanTrainingJobInput,
  EyeTrackingBuleyeanTrainingJobOptions,
  EyeTrackingCalibrationLedger,
  EyeTrackingCalibrationPlan,
  EyeTrackingCalibrationPoint,
  EyeTrackingCalibrationPointStatus,
  EyeTrackingCalibrationRejectionReason,
  EyeTrackingCalibrationSample,
  EyeTrackingCorpus,
  EyeTrackingDatasetSelection,
  EyeTrackingDatasetSlice,
  EyeTrackingDatasetSplit,
  EyeTrackingEyeReading,
  EyeTrackingFailureEvent,
  EyeTrackingFailureKind,
  EyeTrackingFailurePressure,
  EyeTrackingInvariant,
  EyeTrackingMerkleWitness,
  EyeTrackingObservation,
  EyeTrackingRejectionRecord,
  EyeTrackingSession,
  EyeTrackingSessionStatus,
  EyeTrackingSeverity,
  EyeTrackingTrainingManifest,
  EyeTrackingTrainingSplitInput,
  EyeTrackingZone,
  GazePoint,
} from './eyetracking.js';
export {
  GNOSIS_AVATAR_HUMANOID_BONES,
  compileVrmAvatarRuntimeManifest,
  evaluateAvatarPose,
  mixAvatarMorphDeltas,
  sampleProceduralAvatarClip,
  solveLookAtBoneDeltas,
} from './avatar.js';
export type {
  GnosisAvatarAuraState,
  GnosisAvatarBoneDelta,
  GnosisAvatarBoneSlot,
  GnosisAvatarClipName,
  GnosisAvatarCompilationInput,
  GnosisAvatarExpressionProgram,
  GnosisAvatarHumanoidBone,
  GnosisAvatarLookAtProfile,
  GnosisAvatarMorphDelta,
  GnosisAvatarMorphSlot,
  GnosisAvatarMorphSource,
  GnosisAvatarPoseResult,
  GnosisAvatarQuaternion,
  GnosisAvatarRuntimeManifest,
  GnosisAvatarSignalFrame,
  GnosisAvatarVector3,
  GnosisVrmExpressionPreset,
  GnosisVrmExtensionLike,
  GnosisVrmHumanBoneDescriptor,
  GnosisVrmMorphTargetBind,
} from './avatar.js';

// Agent swarm contract surface
export * as agentSwarmContract from './runtime/agent-swarm-contract.js';
export * from './runtime/agent-swarm-contract.js';
export * from './runtime/ambient-scheduler.js';
export * from './runtime/ambient-substrate.js';

// Void primitives — the foundation everything reduces to
export {
  // Boundary
  type VoidBoundary,
  createVoidBoundary,
  updateVoidBoundary,
  decayVoidBoundary,
  boundaryDimensions,
  // Complement distribution
  complementDistribution,
  sampleComplement,
  // Measurement
  type Measurement,
  measure,
  shannonEntropy,
  excessKurtosis,
  giniCoefficient,
  inverseBule,
  // Gait
  type Gait,
  GAIT_DEPTH,
  selectGait,
  // Timescale
  type Timescale,
  TIMESCALE_DECAY,
  type TimescaleBoundary,
  createTimescaleBoundary,
  tickTimescaleBoundary,
  // Boundary stack
  type BoundaryStack,
  createBoundaryStack,
  upwardConstraint,
  downwardContext,
  tickBoundaryStack,
  flattenStack,
  measureStack,
  // Resonance
  type Resonance,
  createResonance,
  // Projection
  projectBoundary,
  // Walker
  type Walker,
  createWalker,
  c0Choose,
  c0Update,
  c1Measure,
  c2c3Adapt,
  // Stack walker
  type StackWalker,
  createStackWalker,
  stepStackWalker,
  measureStackWalker,
} from './void.js';

// Void topology — any GraphAST as a walkable void boundary
export {
  type VoidTopology,
  type ExecutionStep,
  type BranchSelection,
  createVoidTopology,
  rankForkTargets,
  selectActiveBranches,
  sampleSuperposition,
  recordExecution,
  recordEdgeTraversal,
  chooseNext,
  topologyToBoundaryStack,
  measureTopology,
  EDGE_VOID_SEMANTICS,
} from './void-topology.js';

// Void handlers — make void-walking.gg executable
export {
  type VoidWalkingPayload,
  registerVoidWalkingHandlers,
  createVoidWalkingPayload,
} from './void-handlers.js';

// Void agent — METACOG: the agent primitive
export {
  type VoidAgent,
  type VoidAgentConfig,
  type PersonalityLayerConfig,
  type AgentTick,
  createVoidAgent,
  bond,
  perceive,
  perceiveOther,
  decide,
  observe,
  reflect,
  adapt,
  tick,
  completeTick,
  step,
  personalityVector,
  actionPreferences,
  rejectionProfile,
  metacogState,
} from './void-agent.js';

export { QuantumRuntime, type EdgeFabricConfig, type QuantumRuntimeOp, type QuantumRuntimeResult } from './delayed-projection-runtime.js';

export { GnosisNeo4jBridge } from './neo4j-bridge.js';
export { GnosisRegistry } from './runtime/registry.js';
export { GnosisEngine } from './runtime/engine.js';
export { GnosisCoreCache } from './runtime/core-cache.js';
export {
  parseGnarly,
  type GnarlyDocument,
  type GnarlyImplementationBlock,
} from './gnarly-parser.js';
export {
  analyzeGnarly,
  compileGnarly,
  type GnarlyCompileOptions,
  type GnarlyCompileResult,
  type GnarlyDiagnostic,
  type GnarlyGeneratedFile,
  type GnarlySpeedDiagnostic,
} from './gnarly-compiler.js';
export * from './lucas-bridge.js';
export {
  compileTopology,
  executeCompiled,
  executeCompiledSyncFork,
  executeCompiledSync,
  compileSyncForkTopology,
  compileSyncTopology,
  codegenSyncExecutor,
  CannonLauncher,
} from './runtime/compiled-topology.js';
export type {
  CompiledTopology,
  CompiledSyncForkTopology,
  CompiledSyncTopology,
  CompiledStep,
  CompiledSyncStep,
  CompiledFork,
  CompiledExecutionResult,
  SyncHandler,
} from './runtime/compiled-topology.js';
export {
  GnosisWorkerRuntime,
  loadGnosisWorkerRuntime,
} from './runtime/worker-runtime.js';
export type {
  GnosisWorkerExecutionResult,
  GnosisWorkerRuntimeLoadResult,
  GnosisWorkerRuntimeOptions,
} from './runtime/worker-runtime.js';
export {
  coil,
  release,
  getCrossing,
  setCrossing,
  composeKnots,
  hammingDistance,
  toBytes,
  fromBytes,
  edgeTypeToCrossing,
  crossingName,
  VENT,
  FOLD,
  RACE,
  FORK,
  type PackedKnot,
  type CrossingOp,
  type CrossingYield,
  sliceKnot,
  crossover,
  crossover2,
  mutate,
  tournamentSelect,
  evolveGeneration,
  writebackToVFS,
  loadFromVFS,
  evolveAndWriteback,
  type FitnessFunction,
} from './runtime/packed-knot.js';

export * as AmplituhedronKnot from './runtime/amplituhedron-knot.js';

export { BettyCompiler } from './betty/compiler.js';
export {
  compileLegaleseToGg,
  lowerLegaleseToGg,
  segmentLegalese,
} from './betty/legalese.js';
export { compileToGrassmannian, executeGrassmannian } from './betty/grassmannian-compiler.js';
export * from './betty/gg-grassmannian-cache.js';
export { encodeHolographicData, inflateHolographicData } from './betty/holographic-compressor.js';
export {
  encodeCsvHologram,
  inflateCsvHologram,
  encodeJsonArrayHologram,
  inflateJsonArrayHologram,
  encodeMarkdownHologram,
  inflateMarkdownHologram
} from './betty/data-hologram.js';
export { HologramNavigator } from './betty/hologram-navigator.js';
export { executeHologramQuery } from './betty/hologram-query.js';
export { generateLeanFromGnosisAst } from './betty/lean.js';
export { nativeBettiTranspile } from './ts-bridge.js';
export {
  compileLilith,
  compileLilithAsync,
  loadLilith,
  detectBackend,
  detectBackendAsync,
  type LilithResult,
  type LilithInstance,
  type LilithBackend,
} from './lilith.js';
export {
  GnosisModuleLoader,
  createFilesystemModuleResolver,
  detectModuleFormat,
  loadGnosisModuleFromFile,
  looksLikeModule,
  parseMGG,
  renderGraphAst,
  renderMGG,
} from './mod/loader.js';
export {
  analyzeGnosisSource,
  classifySteeringBias,
  createEmptySteeringTelemetry,
  deriveBoundaryWalkBias,
  formatGnosisViolations,
  DEFAULT_GNOSIS_STEERING_MODE,
  finishSteeringTelemetry,
  startSteeringTelemetry,
  VALID_GNOSIS_STEERING_MODES,
  classifySteeringRegime,
  recommendSteeringAction,
  surfaceSteeringMetrics,
  surfaceSteeringRecommendations,
  withSteeringTelemetry,
} from './analysis.js';
export {
  buildBehavioralLoopExecutionTopology,
  buildBehavioralTaxonomyTopology,
  executeBehavioralLoop,
  formatBehavioralTaxonomyMeasurement,
  loadBehavioralLoopsDataset,
  measureBehavioralTaxonomy,
  measureBehavioralTaxonomySelection,
  selectBehavioralTaxonomy,
  writeBehavioralLoopExecutionArtifacts,
} from './behavioral-taxonomy.js';
export {
  authorizeBoundaryWalkRun,
  DEFAULT_GNOSIS_STEERING_TRACE_GUID,
  createSteeringTraceRecord,
  fingerprintSteeringTopology,
  formatSteeringTraceAlert,
  defaultSteeringTraceRoomName,
  formatSteeringTraceStoreStatus,
  formatSteeringTraceSummary,
  GnosisSteeringTraceStore,
  summarizeSteeringTraceRecords,
} from './steering-trace.js';
export {
  createWasmBridgeExecutionAst,
  executeWasmWithGnosis,
  registerWasmBridgeHandlers,
  type ExecuteWasmOptions,
  type WasmExecutionAbi,
  type WasmExecutionHostOptions,
  type WasmExecutionManifest,
  type WasmNodeExecutionPlan,
} from './wasm-execution-bridge.js';
export { GnosisWasmExecutionSurface } from './wasm-execution.js';
export {
  GnosisLeanExecutionSurface,
  LEAN_EXECUTION_CONTRACT,
  LeanExecutorNotRegisteredError,
} from './lean-execution.js';
export {
  GnosisTauriEdgeExecutionSurface,
  buildTauriEdgeReactorPolicy,
  hashTauriEdgeReactorPolicy,
} from './tauri-edge-runtime.js';
export type {
  CacheQueryResolution,
  EscPureSegmentHint,
  EscTier,
  EscWasmSegmentRecord,
  ExecuteWasmArtifactRequest,
  GnosisWasmExecutionSurfaceOptions,
  PositiveFoldWork,
  RegisterWasmArtifactRequest,
  WasmExecutionControlRequest,
  WasmExecutionControlResponse,
  WasmExecutionResponse,
  WasmExecutionSurfaceStatus,
  WasmEscObservation,
} from './wasm-execution.js';
export type {
  ExecuteLeanBundleRequest,
  GnosisLeanExecutionSurfaceOptions,
  LeanExecutionResponse,
  LeanExecutionSurfaceStatus,
  RegisterLeanExecutorRequest,
} from './lean-execution.js';
export type {
  GnosisTauriEdgeExecutionSurfaceOptions,
  TauriEdgeCommandDescriptor,
  TauriEdgeCommandSurfaceDocument,
  TauriEdgeInvokeRequest,
  TauriEdgeInvokeResponse,
  TauriEdgeZkProofEnvelope,
} from './tauri-edge-runtime.js';
export {
  synthesizeAeonFluxSiteWitnesses,
  synthesizeAeonFluxSiteWitnessesFromSource,
  type AeonFluxSiteWitness,
  type AeonFluxSiteAdequacyRegime,
  type AeonFluxSiteAdequacyTheoremRef,
} from './aeon-flux-site-witness.js';
export {
  compileToGGC,
  loadFromGGC,
  isGGC,
  compileSourceToGGC,
  type GGCMetadata,
} from './ggc.js';

export type {
  GnosisHandler,
  GnosisHandlerContext,
} from './runtime/registry.js';
export type {
  GnosisEngineExecutionCacheState,
  GnosisEngineExecutionResult,
  GnosisEngineExecuteOptions,
  GnosisEngineOptions,
} from './runtime/engine.js';
export type {
  GnosisCoreCacheMode,
  GnosisCoreCacheReuseScope,
  GnosisCoreCacheExecutionOptions,
  GnosisCoreCacheSession,
  GnosisCoreCacheMetrics,
  GnosisCoreCacheEntry,
  GnosisCoreCacheCorridorRecord,
  GnosisCoreCacheEvent,
  GnosisCoreCacheLookup,
} from './runtime/core-cache.js';
export type {
  GraphAST,
  ASTNode,
  ASTEdge,
  BettyParseResult,
  BettyLegaleseParseResult,
  Diagnostic,
  DiagnosticCode,
} from './betty/compiler.js';
export type {
  LegaleseActor,
  LegaleseClauseKind,
  LegaleseClauseLowering,
  LegaleseClauseSignal,
  LegaleseClarity,
  LegaleseLoweringOptions,
  LegaleseLoweringResult,
  LegaleseLoweringSummary,
  LegaleseRiskKind,
} from './betty/legalese.js';
export type {
  StabilityContinuousHarrisWitness,
  StabilityContinuousObservableKind,
  StabilityHeteroMoAFabricLayerKind,
  StabilityHeteroMoAFabricLayerWitness,
  StabilityHeteroMoAFabricWitness,
  StabilityKernelEdge,
  StabilityMetadata,
  StabilityProofAssumption,
  StabilityProofKind,
  StabilityProofObligation,
  StabilityRecurrenceStep,
  StabilityRecurrenceWitness,
  StabilityReport,
  StabilityStateAssessment,
} from './betty/stability.js';
export type {
  GnosisModuleImport,
  GnosisModuleExport,
  ParsedGnosisModule,
  LoadedGnosisModule,
  GnosisResolvedImport,
  ResolveResult,
  GnosisModuleResolver,
} from './mod/loader.js';
export type {
  GnosisAnalyzeOptions,
  GnosisCapabilityReport,
  GnosisCollapseCostAction,
  GnosisComplexityReport,
  GnosisFailureTrilemmaMetrics,
  GnosisLineMetrics,
  GnosisQuantumMetrics,
  GnosisSteeringAction,
  GnosisSteeringBias,
  GnosisSteeringEda,
  GnosisSteeringMetrics,
  GnosisSteeringMode,
  GnosisSteeringRegime,
  GnosisSteeringStopwatch,
  GnosisSteeringTelemetry,
  GnosisTopologyMetrics,
} from './analysis.js';
export type {
  BehavioralExecutionContext,
  BehavioralExecutionInterventionKind,
  BehavioralExecutionMode,
  BehavioralExecutionPayload,
  BehavioralExecutionStatus,
  BehavioralLoopArtifactManifest,
  BehavioralLoopArtifactManifestEntry,
  BehavioralLoopArtifactWriteOptions,
  BehavioralLoopCategoryRecord,
  BehavioralLoopExecution,
  BehavioralLoopIntervention,
  BehavioralLoopLogic,
  BehavioralLoopMeta,
  BehavioralLoopOperator,
  BehavioralLoopRecord,
  BehavioralLoopTaxonomy,
  BehavioralLoopsDataset,
  BehavioralTaxonomyMeasureOptions,
  BehavioralTaxonomyMeasurement,
  BehavioralTaxonomySelection,
  BehavioralTaxonomySelectionOptions,
} from './behavioral-taxonomy.js';
export type {
  CreateGnosisSteeringTraceRecordOptions,
  GnosisSteeringTraceCommand,
  GnosisSteeringTraceDataset,
  GnosisSteeringTraceEnvironment,
  GnosisSteeringTraceOutcome,
  GnosisSteeringTraceRecord,
  GnosisSteeringTraceStoreConfig,
  GnosisSteeringTraceStoreStatus,
  GnosisSteeringTraceSummary,
} from './steering-trace.js';

export { generateTlaFromGnosisSource } from './tla-bridge.js';
export type {
  GnosisTlaBridgeResult,
  GnosisTlaBridgeOptions,
} from './tla-bridge.js';

export * from './capabilities/index.js';
export * from './auth/index.js';

// Quantum CRDT — the only state model
export {
  QCorridor,
  QDoc,
  QMap,
  QArray,
  QText,
  QCounter,
  QDocAeonRelay,
  QDocRelay,
  createQDocAeonRelayJoinEnvelope,
} from './crdt/index.js';
export {
  Doc,
  Map,
  Array,
  Text,
  Map as YMap,
  Array as YArray,
  Text as YText,
  XmlFragment,
  XmlElement,
  UndoManager,
  applyUpdate,
  encodeStateAsUpdate,
  transact,
  encodeStateVector,
  diffUpdate,
  mergeUpdates,
} from './crdt/index.js';
export type {
  QCorridorOptions,
  QCorridorExecutionOptions,
  QCorridorMode,
  QCorridorReuseScope,
  QCorridorSession,
  QCorridorMetrics,
  QCorridorEntry,
  QCorridorRecord,
  QCorridorEvent,
  QCorridorEvidence,
  QCorridorObservation,
  QCorridorObservationRole,
  QCorridorObservationStatus,
  QCorridorLookup,
  QDocOptions,
  QDocDelta,
  QDocDeltaNode,
  QDocDeltaEdge,
  QDocUpdateHandler,
  QDocObserveHandler,
  QDocEvent,
  QMapKeyChange,
  QMapEvent,
  QDocAeonRelayAttributes,
  QDocAeonRelayAttributeValue,
  QDocAeonRelayConfig,
  QDocAeonRelayJoinEnvelope,
  QDocAeonRelayReadyStrategy,
  QDocAeonRelaySpan,
  QDocAeonRelayStatus,
  QDocAeonRelayTelemetry,
  QDocAeonRelayTelemetryEvent,
  QDocAeonRelayTelemetryStage,
  QDocRelayAttributes,
  QDocRelayAttributeValue,
  QDocRelayConfig,
  QDocRelayReadyStrategy,
  QDocRelaySpan,
  QDocRelayStatus,
  QDocRelayTelemetry,
  QDocRelayTelemetryEvent,
  QDocRelayTelemetryStage,
} from './crdt/index.js';

// Test harness & runner
export {
  ggTest,
  ggQuickCheck,
  ggAssert,
  GGTestBuilder,
} from './gg-test-harness.js';
export {
  runGGTestFile,
  runGGTestSuite,
  discoverTestFiles,
  formatGGTestResults,
  formatGGTestDiscoveryResults,
} from './gg-test-runner.js';

// `.gg`-native neural runtime (used by @affectively/neural bridge)
export {
  NeuralEngine,
  GPUEngine,
  WebGPUEngine,
  WebNNEngine,
  Translator,
  NeuronRepository,
  SynapseRepository,
  TOPIC_DOMAIN_TRANSFORMER_TOPOLOGY_FILE,
  TOPIC_DOMAIN_TRANSFORMER_TOPOLOGY,
  TRANSFORMER_HELLO_WORLD_TOPOLOGY,
  getTopicDomainTransformerTopologySource,
  init as initNeuralEngine,
  init as init,
} from './neural-compat.js';

export type {
  Neuron,
  Synapse,
  AdapterTrainingConfig,
  NeuralGraphData,
  LoadTopologyOptions,
  NeuralHeteroFabricOptions,
} from './neural-compat.js';
export * from './runtime/hetero-fabric.js';

export {
  LanguageAdapterRegistry,
  buildCodeToNaturalArtifact,
  createPreservationProfile,
  createSourceArtifactFromFile,
  crossCompile,
  defaultLanguageAdapterRegistry,
  inferCodeLanguageFromFilePath,
  inferLanguageDomainFromFilePath,
  installBuiltInLanguageAdapters,
} from './cross-domain.js';
export type {
  CrossCompileContext,
  CrossCompileRequest,
  CrossCompileResult,
  CrossDomainAnalysis,
  LanguageAdapter,
  LanguageDomain,
  PreservationProfile,
  SourceArtifact,
  TargetArtifact,
  TargetArtifactDescriptor,
} from './cross-domain.js';
export type {
  SemanticFacet,
  SemanticObligation,
} from './betty/semantic-compatibility.js';
export type {
  JsonSchema,
  SemanticContract,
  SemanticPredicate,
  TopologyType,
} from './betty/semantic-compatibility.js';

export {
  GnosisTypeScriptBridgeError,
  compileTypeScriptToGnosis,
  executeTypeScriptWithGnosis,
  registerTypeScriptBridgeHandlers,
  renderTypeScriptBridgeRuntimeModule,
} from './ts-bridge.js';

export type {
  ExecuteTypeScriptWithGnosisOptions,
  GnosisTypeScriptBridgeBinding,
  GnosisTypeScriptBridgeBindings,
  GnosisTypeScriptBridgeErrorLocation,
  GnosisTypeScriptBridgeExpression,
  GnosisTypeScriptBridgeNodePlan,
  GnosisTypeScriptBridgeOptions,
  GnosisTypeScriptBridgeResult,
  GnosisTypeScriptBridgeSourceLocation,
  GnosisTypeScriptBridgeWave,
} from './ts-bridge.js';

export { GnosisGnodeExecutionSurface } from './gnode-execution.js';

export type {
  ExecuteGnodeArtifactRequest,
  GnodeExecutionControlRequest,
  GnodeExecutionControlResponse,
  GnodeExecutionLanePlan,
  GnodeExecutionObservation,
  GnodeExecutionPlan,
  GnodeExecutionProfileSnapshot,
  GnodeExecutionResponse,
  GnodeExecutionStrategy,
  GnodeExecutionSurfaceStatus,
  GnodeExecutionWavePlan,
  GnosisGnodeExecutionSurfaceOptions,
  RegisterGnodeArtifactRequest,
  SerializedGnodeArtifact,
} from './gnode-execution.js';

export { checkTypeScriptWithGnosis } from './ts-check.js';

export type {
  GnosisGhostMassLedger,
  GnosisTypeScriptCheckDiagnostic,
  GnosisTypeScriptCheckDiagnosticLevel,
  GnosisTypeScriptCheckMetrics,
  GnosisTypeScriptCheckOptions,
  GnosisTypeScriptCheckResult,
  GnosisTypeScriptCheckTopologyEdge,
  GnosisTypeScriptCheckTopologyGraph,
  GnosisTypeScriptCheckTopologyNode,
} from './ts-check.js';

// Incremental check with caching
export { GnosisIncrementalChecker } from './ts-check-incremental.js';
export type {
  IncrementalCacheEntry,
  IncrementalCheckStats,
} from './ts-check-incremental.js';

// Autofix engine
export {
  detectDeadBranches,
  generateAutofixSuggestions,
} from './ts-check-autofix.js';
export type {
  AutofixKind,
  AutofixReplacement,
  AutofixSuggestion,
  DeadBranchInfo,
} from './ts-check-autofix.js';

// Cross-function call graph
export { buildCallGraph } from './ts-call-graph.js';
export type {
  CallGraph,
  CallGraphEdge,
  CallGraphExternalImport,
  CallGraphModule,
  CallGraphNode,
  CallGraphOptions,
} from './ts-call-graph.js';

// Topology diff for PR reviews
export { computeTopologyDiff, formatPrComment } from './ts-check-diff.js';
export type {
  FunctionDiff,
  TopologyDiffOptions,
  TopologyDiffResult,
  TopologyMetricsDelta,
  TopologyRegimeTransition,
} from './ts-check-diff.js';

// File watcher for live feedback
export { GnosisFileWatcher } from './ts-check-watcher.js';
export type {
  WatcherEvent,
  WatcherListener,
  WatcherOptions,
} from './ts-check-watcher.js';

// Lean certificate generation
export {
  generateLeanCertificate,
  generateCertificatesForDirectory,
  verifyCertificate,
} from './ts-lean-certificate.js';
export type {
  LeanCertificate,
  LeanCertificateOptions,
} from './ts-lean-certificate.js';

// Topology SLOs
export {
  checkDirectorySlo,
  checkTopologySlo,
  loadSloConfig,
} from './ts-topology-slo.js';
export type {
  SloCheckResult,
  SloViolation,
  TopologySloConfig,
} from './ts-topology-slo.js';

// Runtime topology probes
export {
  GnosisProbeCollector,
  createProbeWrapper,
  buildOtelAttributes,
} from './ts-runtime-probe.js';
export type {
  BranchFrequency,
  NodeHotspot,
  OtelSpanAttributes,
  ProbeEvent,
  ProbeSession,
  RuntimeTopologyMetrics,
  WallaceDivergence,
} from './ts-runtime-probe.js';

// Seven Laws & God Formula check system
export {
  checkSevenLaws,
  checkGodFormula,
  checkControlStatistics,
  checkReductionChain,
  fullBuleyeanAudit,
  auditVoidBoundary,
  formatAuditReport,
  type CheckResult,
  type SevenLawsReport,
  type GodFormulaReport,
  type ControlStatisticReport,
  type ReductionChainReport,
  type FullAuditReport,
  type BuleyeanInput,
} from './seven-laws-checker.js';

// Formal ledger kernel
export * from './void-dimensions.js';
export * from './flow-obstruction.js';
export * from './reduction-projection-scan.js';
export * from './reduction-projection-formalization.js';
export * from './decomposition-topology.js';
export * from './transformation-family-formalization.js';
export * from './juggling-orbit-formalization.js';
export * from './certified-allocation-kernel.js';
export * from './advisory-allocation-kernel.js';
export * from './hft-advisory-guard.js';
export * from './juggling-chaos-orchestration.js';
export * from './pattern-horizon-compiler.js';
export * from './sofa-topology.js';
export * from './formal-ledger.js';
export * from './twin-psyche-formalization.js';
export * from './spiderweb-topology.js';
export * from './triangulation-dimensional-coercion.js';

// Bitwise Cypher (Lacey Variant)
export {
  Xorshift64,
  encodeHelixStreamId,
  decodeHelixStreamId,
  type HelixCipherKey,
  helixKeyFromSeed,
  helixEncryptFrame,
  helixDecryptFrame,
  helixEncryptStream,
  helixDecryptStream,
  HELIX_TURN_SAMPLES,
  type RotatingHelixKey,
  rotatingHelixKeyFromSeed,
  rotationAt,
  isAlphaHalf,
  rotatingHelixEncryptFrame,
  rotatingHelixDecryptFrame,
  rotatingHelixEncryptStream,
  rotatingHelixDecryptStream,
  seedFromPassphrase,
  syzygySmush,
  syzygyUnsmush,
  syzygyFoldBinary,
  syzygyUnfoldBinary,
  laceyEncryptFrame,
  laceyDecryptFrame,
  laceyEncryptStream,
  laceyDecryptStream,
} from './bitwise-cipher.js';

export * from './aperiodic-scheduling.js';
export * from './randomness.js';
export * from './topological-xor-certificate.js';
export * from './physics-sandbox.js';
export * from './seismic-stress-toy.js';
export * from './math-sandbox.js';
export * from './wallington-rotation.js';
export * from './god-formula-vm.js';
export * from './unspeech-triton.js';
export * from './unspeech-unicode-teleport.js';
export { assertThermodynamicClearing } from './holy-trinity.js';
export { analyzeConversionTopology, type ConversionTopologyResult } from './conversion-tracking-topology.js';
export { evaluateViewportPacking, type ViewportPackingResult } from './knapsack-viewport-topology.js';
export * from './scheduling-mesh.js';
export * from './dark-fork-topology.js';
export * from './foil-edge-gate.js';

// ============================================================================
// Betti Formal Ledger Bridges (Lean-verified, zero sorry)
// ============================================================================
export * from './betti/moonshine-trace.js';
export * from './betti/moonshine-router.js';
export * from './betti/topological-memoization.js';
export * from './betti/pisot-stabilized-intelligence.js';
export * from './betti/retrocausal-memoization.js';
export * from './betti/skyrms-syzygy-equilibrium.js';
export * from './betti/genesis-query.js';
export * from './betti/buley-equilibrium.js';
