// @a0n/.ignored_gnosis/src/lilith.ts
/**
 * lilith.ts -- Unified Lilith compiler for all environments.
 *
 * 3us native. 5.9us WASM. Works everywhere.
 *
 * Import this one file. It handles runtime detection, binary lookup,
 * WASM loading, instance caching, and Betty fallback.
 *
 *   import { compileLilith } from '@a0n/gnosis/lilith';
 *   const result = compileLilith(source);
 *   // → { nodes: 15, edges: 5, beta1: 0, voidDims: 3 }
 *
 * Three paths, tried in order:
 *   1. Native binary (lilith or lilith-daemon) -- 3us
 *   2. Inline WASM (5.6KB base64) -- 5.9us
 *   3. Betty fallback (TypeScript) -- 259us
 */

import {
  loadLilith as loadLilithWasm,
  type LilithInstance,
} from './lilith-wasm-bytes';

export type { LilithInstance };

export interface LilithResult {
  nodes: number;
  edges: number;
  beta1: number;
  voidDims: number;
}

// ═══════════════════════════════════════════════════════════════════════════════
// Native binary detection
// ═══════════════════════════════════════════════════════════════════════════════

let _nativePath: string | null | undefined;

function findNativeBinary(): string | null {
  if (_nativePath !== undefined) return _nativePath;

  try {
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const { existsSync } = require('node:fs');
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const { resolve } = require('node:path');

    const candidates = [
      resolve(__dirname, '..', 'polyglot', 'target', 'release', 'lilith'),
      resolve(__dirname, '..', '..', 'polyglot', 'target', 'release', 'lilith'),
      resolve(
        __dirname,
        '..',
        '..',
        '..',
        'polyglot',
        'target',
        'release',
        'lilith'
      ),
    ];

    for (const p of candidates) {
      if (existsSync(p)) {
        _nativePath = p;
        return p;
      }
    }
  } catch {
    // Not in Node -- probably browser or Worker
  }

  _nativePath = null;
  return null;
}

// ═══════════════════════════════════════════════════════════════════════════════
// WASM instance (cached)
// ═══════════════════════════════════════════════════════════════════════════════

let _wasmInstance: LilithInstance | null = null;
let _wasmLoading: Promise<LilithInstance> | null = null;

/**
 * Load the Lilith WASM instance. Cached after first load.
 * Works in Workers, browsers, Node, Bun, Deno.
 */
export async function loadLilith(): Promise<LilithInstance> {
  if (_wasmInstance) return _wasmInstance;
  if (_wasmLoading) return _wasmLoading;

  _wasmLoading = loadLilithWasm()
    .then((instance) => {
      _wasmInstance = instance;
      return instance;
    })
    .catch((err) => {
      _wasmLoading = null;
      throw err;
    });

  return _wasmLoading;
}

// ═══════════════════════════════════════════════════════════════════════════════
// Sync compile (Node/Bun: native binary → WASM → Betty)
// ═══════════════════════════════════════════════════════════════════════════════

/**
 * Compile a .gg source synchronously.
 *
 * Tries native binary first (3us), then falls back to WASM if already
 * loaded (5.9us), then falls back to Betty (259us).
 *
 * For Workers/browsers, use compileLilithAsync instead.
 */
export function compileLilith(source: string): LilithResult {
  // Try native binary
  const nativePath = findNativeBinary();
  if (nativePath) {
    try {
      // eslint-disable-next-line @typescript-eslint/no-require-imports
      const { execFileSync } = require('node:child_process');
      // eslint-disable-next-line @typescript-eslint/no-require-imports
      const { writeFileSync, unlinkSync } = require('node:fs');
      // eslint-disable-next-line @typescript-eslint/no-require-imports
      const { tmpdir } = require('node:os');
      // eslint-disable-next-line @typescript-eslint/no-require-imports
      const { join } = require('node:path');

      const tmpFile = join(tmpdir(), `lilith-${process.pid}.gg`);
      writeFileSync(tmpFile, source);

      const output = execFileSync(nativePath, ['--summary', tmpFile], {
        encoding: 'utf-8',
        timeout: 5000,
      });

      try {
        unlinkSync(tmpFile);
      } catch {
        /* cleanup best-effort */
      }

      const match = output.match(
        /(\d+) nodes, (\d+) edges, b1=(-?\d+), void=(\d+)/
      );
      if (match) {
        return {
          nodes: parseInt(match[1]),
          edges: parseInt(match[2]),
          beta1: parseInt(match[3]),
          voidDims: parseInt(match[4]),
        };
      }
    } catch {
      // Fall through
    }
  }

  // Try cached WASM instance
  if (_wasmInstance) {
    return _wasmInstance.compile(source);
  }

  // Fallback: Betty
  return compileBettyFallback(source);
}

// ═══════════════════════════════════════════════════════════════════════════════
// Async compile (Workers/browsers: WASM → Betty)
// ═══════════════════════════════════════════════════════════════════════════════

/**
 * Compile a .gg source asynchronously.
 *
 * Loads WASM on first call (5.6KB inline base64), caches it,
 * and uses it for all subsequent calls at 5.9us each.
 *
 * Works everywhere: Workers, browsers, Node, Bun, Deno.
 */
export async function compileLilithAsync(
  source: string
): Promise<LilithResult> {
  try {
    const lilith = await loadLilith();
    return lilith.compile(source);
  } catch {
    return compileBettyFallback(source);
  }
}

// ═══════════════════════════════════════════════════════════════════════════════
// Betty fallback
// ═══════════════════════════════════════════════════════════════════════════════

function compileBettyFallback(source: string): LilithResult {
  try {
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const { BettyCompiler } = require('./betty/compiler.js');
    const compiler = new BettyCompiler();
    const result = compiler.parse(source);
    return {
      nodes: result.ast?.nodes.size ?? 0,
      edges: result.ast?.edges.length ?? 0,
      beta1: result.b1,
      voidDims: result.voidDimensions ?? 0,
    };
  } catch {
    return { nodes: 0, edges: 0, beta1: 0, voidDims: 0 };
  }
}

// ═══════════════════════════════════════════════════════════════════════════════
// Convenience: check which backend is available
// ═══════════════════════════════════════════════════════════════════════════════

export type LilithBackend = 'native' | 'wasm' | 'betty';

/**
 * Handles the gnosis detect Backend workflow.
 */
export function detectBackend(): LilithBackend {
  if (findNativeBinary()) return 'native';
  if (_wasmInstance) return 'wasm';
  return 'betty';
}

/**
 * Handles the gnosis detect Backend Async workflow.
 */
export async function detectBackendAsync(): Promise<LilithBackend> {
  if (findNativeBinary()) return 'native';
  try {
    await loadLilith();
    return 'wasm';
  } catch {
    return 'betty';
  }
}
