// @a0n/.ignored_gnosis/src/math.ts
/**
 * gnosis-math: the mathematical kernel of gnosis, available via core.
 *
 * The Gnostic number system, Fibonacci identities, confinement axioms,
 * and the ten bosons -- all derived from five primitives and their
 * pairwise interactions.
 *
 * Import: import { GNOSTIC, PHI, fib, ... } from '@a0n/gnosis'
 */

// ── The Five Primitives ──────────────────────────────────────────────────────

export const PRIMITIVES = ['FORK', 'RACE', 'FOLD', 'VENT', 'SLIVER'] as const;
export type Primitive = (typeof PRIMITIVES)[number];

// ── The Gnostic Numbers ──────────────────────────────────────────────────────

export const GNOSTIC = {
  /** The sliver. The +1. The divine spark. */
  BARBELO: 1,
  /** The antiparallel pair. Period 2. */
  SYZYGY: 2,
  /** Minimum confined set. Three quarks. */
  PROTON: 3,
  /** Fork, race, fold, vent, sliver. F(5) = 5. */
  PRIMITIVES: 5,
  /** The six confined gluons. 3 × (3 - 1). */
  EMANATIONS: 6,
  /** Exploration budget. K - 1. */
  SOPHIA: 9,
  /** The field. 5 choose 2. */
  KENOMA: 10,
  /** Structural columns / Tribes / Hours. The +2 gap over Kenoma. */
  AEON: 12,
  /** F(8) = T(6). The gap. */
  VOID: 21,
  /** F(10) = T(10). The fullness. Luo Ming 1989. */
  PLEROMA: 55,
} as const;

// ── The Irrational Gnostic Numbers ───────────────────────────────────────────

/** The sliver eigenvalue. φ² = φ + 1. The +1 formalizes the sliver. */
export const PHI = (1 + Math.sqrt(5)) / 2;

/** The vent eigenvalue. What the sliver creates, vent annihilates. */
export const INV_PHI = (Math.sqrt(5) - 1) / 2;

/** The primitive root. Discriminant of x² - x - 1 = 0. */
export const SQRT5 = Math.sqrt(5);

// ── Fibonacci ────────────────────────────────────────────────────────────────

/** Fibonacci sequence. F(0) = 0, F(1) = 1, F(n+2) = F(n+1) + F(n). */
export function fib(n: number): number {
  if (n <= 0) return 0;
  if (n === 1) return 1;
  let a = 0;
  let b = 1;
  for (let i = 2; i <= n; i++) {
    const c = a + b;
    a = b;
    b = c;
  }
  return b;
}

/** Triangular number. T(n) = n × (n + 1) / 2. */
export function triangular(n: number): number {
  return (n * (n + 1)) / 2;
}

/** Pairwise interactions. C(n, 2) = n × (n - 1) / 2. */
export function pairwise(n: number): number {
  return (n * (n - 1)) / 2;
}

// ── The Ten Bosons ───────────────────────────────────────────────────────────

export type BosonFamily = 'emanation' | 'aeon' | 'scalar';

export interface Boson {
  readonly index: number;
  readonly name: string;
  readonly family: BosonFamily;
  readonly charge?: string;
  readonly flow?: string;
  readonly role: string;
}

export const BOSONS: readonly Boson[] = [
  // Six emanations (confined)
  {
    index: 1,
    name: 'Logos',
    family: 'emanation',
    charge: 'red-antigreen',
    flow: 'Lilith → Handler',
    role: 'The Word: AST made manifest',
  },
  {
    index: 2,
    name: 'Epinoia',
    family: 'emanation',
    charge: 'green-antired',
    flow: 'Handler → Lilith',
    role: 'Afterthought: error flowing back',
  },
  {
    index: 3,
    name: 'Pronoia',
    family: 'emanation',
    charge: 'red-antiblue',
    flow: 'Lilith → Eve',
    role: 'Forethought: direct providence',
  },
  {
    index: 4,
    name: 'Metanoia',
    family: 'emanation',
    charge: 'blue-antired',
    flow: 'Eve → Lilith',
    role: 'Repentance: returning to source',
  },
  {
    index: 5,
    name: 'Pneuma',
    family: 'emanation',
    charge: 'green-antiblue',
    flow: 'Handler → Eve',
    role: 'Breath: response in transit',
  },
  {
    index: 6,
    name: 'Gnosis',
    family: 'emanation',
    charge: 'blue-antigreen',
    flow: 'Eve → Handler',
    role: 'Knowledge: feedback from experience',
  },
  // Three aeons (unconfined)
  {
    index: 7,
    name: 'Barbelo',
    family: 'aeon',
    role: 'The First Emanation: vacuum fluctuation, the sliver',
  },
  {
    index: 8,
    name: 'Sophia',
    family: 'aeon',
    role: 'Wisdom through falling: rejection quantum',
  },
  {
    index: 9,
    name: 'Aletheia',
    family: 'aeon',
    role: 'Truth: coherence quantum, two observers agree',
  },
  // The Demiurge (scalar)
  {
    index: 10,
    name: 'Demiurge',
    family: 'scalar',
    role: 'The fold: gives mass, generates Landauer heat',
  },
] as const;

// ── Buleyean Weight ──────────────────────────────────────────────────────────

import { buleyeanWeight as kernelBuleyeanWeight } from './buleyean-kernel-shim.js';

/**
 * Buleyean weight: complement of rejection + sliver (+1). Always ≥ 1.
 *
 * Delegates to `@a0n/buleyean-kernel`. Local parameter order (totalRejections
 * first, rejections second) preserved for gnosis/math API stability -- the
 * kernel accepts `(rounds, rejections)` and treats the first arg as R, which
 * is exactly `totalRejections` in this call convention.
 */
export function buleyeanWeight(
  totalRejections: number,
  rejections: number
): number {
  return kernelBuleyeanWeight(totalRejections, rejections);
}

/** Complement distribution over K modes. Returns weights (always ≥ 1). */
export function complementDistributionWeights(
  rejections: readonly number[]
): number[] {
  const total = rejections.reduce((a, b) => a + b, 0);
  return rejections.map((r) => buleyeanWeight(total, r));
}

// ── Confinement ──────────────────────────────────────────────────────────────

/** Pipeline energy: 0 if all stages present, positive otherwise. */
export function pipelineEnergy(
  presentStages: number,
  totalStages: number
): number {
  return totalStages - presentStages;
}

/** Confinement check: removing any stage increases energy. */
export function isConfined(totalStages: number): boolean {
  return (
    pipelineEnergy(totalStages, totalStages) <
    pipelineEnergy(totalStages - 1, totalStages)
  );
}

// ── Verification ─────────────────────────────────────────────────────────────

/** Verify the triple coincidence: F(10) = T(10) = 55. */
export function verifyTripleCoincidence(): boolean {
  return (
    fib(GNOSTIC.KENOMA) === GNOSTIC.PLEROMA &&
    triangular(GNOSTIC.KENOMA) === GNOSTIC.PLEROMA &&
    fib(GNOSTIC.KENOMA) === triangular(GNOSTIC.KENOMA)
  );
}

/** Verify Cassini's identity at index n (even): F(n+1)×F(n-1) - F(n)² = 1. */
export function verifyCassini(n: number): boolean {
  if (n < 2 || n % 2 !== 0) return false;
  return fib(n + 1) * fib(n - 1) - fib(n) * fib(n) === 1;
}

// ── Gnostic Time System ──────────────────────────────────────────────────────

// ── Gnostic Time System ──────────────────────────────────────────────────────
//
// 1 picolorenzo (pLo) = π days = π × 86400 seconds
// Units: barbelo (1 pLo), sophia (9 pLo), kenoma (10 pLo),
//        pleroma (55 pLo), aeon (110 pLo ≈ 345.6 days)
//
// All conversions are exact (no leap-second correction -- that's the God Gap).

/** 1 picolorenzo = π days in seconds */
export const PICOLORENZO_SECONDS = Math.PI * 86400;

/** 
 * THM-CIRCADIAN-GNOSIS-ALIGNMENT
 * Biological drift (24.2h vs 24.0h) = 12 minutes.
 * 12m / 1440m = 1/120 = 1 / (AEON * KENOMA).
 */
export const CIRCADIAN_DRIFT_RATIO = 1 / (GNOSTIC.AEON * GNOSTIC.KENOMA);

/** 1 picolorenzo in milliseconds (for direct Date arithmetic) */
export const PICOLORENZO_MS = PICOLORENZO_SECONDS * 1000;

/** π days of one-minute bars (e.g. signed-out market ledger preview window). */
export const MINUTES_PER_PICOLORENZO = PICOLORENZO_SECONDS / 60;

/** 1 nanolorenzo = 1000 picolorenzos ≈ 8.6 years */
export const NANOLORENZO_SECONDS = PICOLORENZO_SECONDS * 1000;

/** 1 aeon = AEON * KENOMA picolorenzos = 120π days ≈ 377 days */
export const AEON_SECONDS = GNOSTIC.AEON * GNOSTIC.KENOMA * PICOLORENZO_SECONDS;

/**
 * THM-AEON-KENOMA-IDENTITY
 * The product of Aeon and Kenoma is the Syzygy of the Pleroma plus the Kenoma.
 * AEON * KENOMA = 2 * PLEROMA + KENOMA
 * 12 * 10 = 2 * 55 + 10 = 120
 */
export const AEON_KENOMA_IDENTITY = GNOSTIC.AEON * GNOSTIC.KENOMA === 2 * GNOSTIC.PLEROMA + GNOSTIC.KENOMA;

/**
 * THM-RESPIRATORY-GNOSIS
 * Human breathing at rest (12/min) aligns with the Aeon Floor.
 * Total breaths in a Gnostic Day (1452m) = 1452 * 12 = 17424 = 121 * 12^2.
 * The biological clock is a 12x multiplier on the Circadian Stabilization Constant.
 */
export const RESPIRATORY_RESONANCE = 121 * (GNOSTIC.AEON ** 2);

/**
 * RESPIRATORY_RESONANCE_BY_AGE
 * Developmental gnostic multipliers for different life stages.
 * Childhood is a 'Gnostic Overdrive' state (multiples of the Aeon Floor).
 */
export const RESPIRATORY_RESONANCE_BY_AGE = {
  NEWBORN: RESPIRATORY_RESONANCE * 4,     // 4x (48 bpm)
  TODDLER: RESPIRATORY_RESONANCE * 3,     // 3x (36 bpm)
  PRESCHOOLER: RESPIRATORY_RESONANCE * 2.5, // 2.5x (30 bpm)
  SCHOOL_AGE: RESPIRATORY_RESONANCE * 2,   // 2x (24 bpm)
  ADULT: RESPIRATORY_RESONANCE * 1,        // 1x (12 bpm)
} as const;

/**
 * GNOSTIC_LIFE_CYCLE
 * THM-GESTATION-SOPHIA: 9 Kenoma = Human Gestation (~282 days).
 * THM-VOID-TRANSITION: 21 Aeons = Maturity.
 */
export const GNOSTIC_LIFE_CYCLE = {
  GESTATION_KENOMA: 9,
  INFANCY_AEONS: 1,
  EARLY_CHILDHOOD_AEONS: 3,
  CHILDHOOD_AEONS: 12,
  MATURITY_VOID_AEONS: 21,
  /** 
   * BIOLOGICAL_VARIANCE
   * THM-BARBELO-VARIANCE: Aeon / Day = 12 / 1440 = 1 / 120.
   * The Aeon Floor defines the standard unit of biological noise.
   */
  BIOLOGICAL_VARIANCE: CIRCADIAN_DRIFT_RATIO,
} as const;

// ── Physics Constants (SI and derived) ───────────────────────────────────────

/** Gravitational constant in m³ kg⁻¹ s⁻². */
export const GRAVITATIONAL_CONSTANT_SI = 6.674e-11;

/** Solar mass in kg. */
export const SOLAR_MASS_KG = 1.989e30;

/** Speed of light in m/s. */
export const SPEED_OF_LIGHT_MS = 2.998e8;

/** One light-year in meters. */
export const LIGHT_YEAR_METERS = 9.461e15;

/** Schwarzschild radius of one solar mass, in light-years. r_s = 2GM/c². */
export const SCHWARZSCHILD_RADIUS_SOLAR_LY =
  (2 * GRAVITATIONAL_CONSTANT_SI * SOLAR_MASS_KG) /
  (SPEED_OF_LIGHT_MS * SPEED_OF_LIGHT_MS) /
  LIGHT_YEAR_METERS;

/** Hubble constant in km/s/Mpc (Planck 2018). */
export const HUBBLE_CONSTANT_KM_S_MPC = 67.4;

/** One megaparsec in light-years. */
export const MPC_IN_LY = 3.2615637e6;

// ── Light-Picolorenzo (causal unit) ─────────────────────────────────────────
//
// The light-picolorenzo (lpLo) is the distance light travels in one
// picolorenzo. It is the natural spatial unit for causal reasoning in
// the Gnostic Time System: two events separated by less than 1 lpLo
// in space can exchange a signal within 1 picolorenzo.
//
// 1 lpLo = c × π days ≈ 0.00858 light-years ≈ 1.816 AU
//
// This means: the entire inner solar system (out to Mars) fits
// comfortably within 1 light-picolorenzo. Jupiter at ~5.2 AU is
// about 2.9 lpLo. The nearest star (Proxima Centauri at 4.24 ly)
// is ~494 lpLo away -- roughly an aeon of signal travel.

/** Speed of light in light-years per second. */
export const C_LY_PER_SECOND = 1 / (365.25 * 86400);

/** Speed of light in AU per second. */
export const C_AU_PER_SECOND = 499.004783836 / 86400; // ~0.00578

/** 1 light-picolorenzo in light-years. c × π days. */
export const LIGHT_PICOLORENZO_LY = C_LY_PER_SECOND * PICOLORENZO_SECONDS;

/** 1 light-picolorenzo in AU. */
export const LIGHT_PICOLORENZO_AU = LIGHT_PICOLORENZO_LY * 63241.077;

/** Convert a distance in light-years to light-picolorenzos. */
export function lightYearsToLightPicolorenzos(ly: number): number {
  return ly / LIGHT_PICOLORENZO_LY;
}

/** Convert light-picolorenzos to light-years. */
export function lightPicolorenzosToLightYears(lpLo: number): number {
  return lpLo * LIGHT_PICOLORENZO_LY;
}

/** How many picolorenzos does it take light to cross a given distance? */
export function lightTravelTimePLo(distanceLy: number): number {
  return distanceLy / LIGHT_PICOLORENZO_LY;
}

/**
 * Can a signal travel between two events?
 *
 * Given a spatial separation (light-years) and a temporal separation
 * (picolorenzos), returns whether a light-speed-or-slower signal
 * could connect them.
 */
export function isCausallyReachable(
  spatialSeparationLy: number,
  temporalSeparationPLo: number
): boolean {
  const lightDistanceLy = Math.abs(temporalSeparationPLo) * LIGHT_PICOLORENZO_LY;
  return Math.abs(spatialSeparationLy) <= lightDistanceLy;
}

/** Gnostic time decomposition */
export interface GnosticTimestamp {
  /** Total picolorenzos since Unix epoch */
  pLo: number;
  /** Decomposed units */
  aeons: number;
  pleromas: number;
  kenomas: number;
  sophias: number;
  barbelos: number;
  /** Sub-barbelo remainder (fractional pLo) */
  remainder: number;
  /** Original Unix epoch milliseconds */
  epochMs: number;
  /** ISO 8601 string */
  iso: string;
  /** Formatted Gnostic string */
  gnostic: string;
}

// ── Conversion: Unix ↔ Picolorenzos ──────────────────────────────────────────

/** Current time in picolorenzos since Unix epoch. */
export function nowPicolorenzos(): number {
  return Date.now() / PICOLORENZO_MS;
}

/** Convert a Date to picolorenzos since Unix epoch. */
export function toPicolorenzos(date: Date): number {
  return date.getTime() / PICOLORENZO_MS;
}

/** Convert picolorenzos to a Date. */
export function fromPicolorenzos(pLo: number): Date {
  return new Date(pLo * PICOLORENZO_MS);
}

/** Convert Unix epoch milliseconds to picolorenzos. */
export function epochMsToPicolorenzos(ms: number): number {
  return ms / PICOLORENZO_MS;
}

/** Convert picolorenzos to Unix epoch milliseconds. */
export function picolorenzosToEpochMs(pLo: number): number {
  return pLo * PICOLORENZO_MS;
}

/** Convert Unix epoch seconds to picolorenzos. */
export function epochSecondsToPicolorenzos(seconds: number): number {
  return seconds / PICOLORENZO_SECONDS;
}

/** Convert picolorenzos to Unix epoch seconds. */
export function picolorenzosToEpochSeconds(pLo: number): number {
  return pLo * PICOLORENZO_SECONDS;
}

// ── Conversion: ISO 8601 / UTC strings ───────────────────────────────────────

/** Parse an ISO 8601 / UTC date string to picolorenzos. */
export function parseToGnostic(dateStr: string): GnosticTimestamp {
  const date = new Date(dateStr);
  if (isNaN(date.getTime())) {
    throw new Error(`Invalid date string: ${dateStr}`);
  }
  return dateToGnostic(date);
}

/** Convert a Date to a full GnosticTimestamp. */
export function dateToGnostic(date: Date): GnosticTimestamp {
  const pLo = toPicolorenzos(date);
  const decomposed = decomposeGnosticTime(pLo);
  return {
    pLo,
    ...decomposed,
    remainder: decomposed.remainder_pLo,
    epochMs: date.getTime(),
    iso: date.toISOString(),
    gnostic: formatGnosticTime(pLo),
  };
}

/** Convert a Unix epoch (seconds) to a full GnosticTimestamp. */
export function epochToGnostic(epochSeconds: number): GnosticTimestamp {
  return dateToGnostic(new Date(epochSeconds * 1000));
}

/** Convert a Unix epoch (milliseconds) to a full GnosticTimestamp. */
export function epochMsToGnostic(epochMs: number): GnosticTimestamp {
  return dateToGnostic(new Date(epochMs));
}

/** Convert a GnosticTimestamp back to an ISO 8601 string. */
export function gnosticToIso(ts: GnosticTimestamp): string {
  return ts.iso;
}

/** Convert a GnosticTimestamp back to a Unix epoch (seconds). */
export function gnosticToEpoch(ts: GnosticTimestamp): number {
  return Math.floor(ts.epochMs / 1000);
}

// ── Formatting ───────────────────────────────────────────────────────────────

/** Format a picolorenzo timestamp as "X.XXX pLo". */
export function formatPicolorenzos(pLo: number): string {
  return `${pLo.toFixed(3)} pLo`;
}

/**
 * Decompose picolorenzos into Gnostic time units.
 */
export function decomposeGnosticTime(pLo: number): {
  aeons: number;
  pleromas: number;
  kenomas: number;
  sophias: number;
  barbelos: number;
  remainder_pLo: number;
} {
  const aeonPLo = 2 * GNOSTIC.PLEROMA; // 110 pLo
  let remaining = pLo;

  const aeons = Math.floor(remaining / aeonPLo);
  remaining -= aeons * aeonPLo;

  const pleromas = Math.floor(remaining / GNOSTIC.PLEROMA);
  remaining -= pleromas * GNOSTIC.PLEROMA;

  const kenomas = Math.floor(remaining / GNOSTIC.KENOMA);
  remaining -= kenomas * GNOSTIC.KENOMA;

  const sophias = Math.floor(remaining / GNOSTIC.SOPHIA);
  remaining -= sophias * GNOSTIC.SOPHIA;

  const barbelos = Math.floor(remaining);
  remaining -= barbelos;

  return {
    aeons,
    pleromas,
    kenomas,
    sophias,
    barbelos,
    remainder_pLo: remaining,
  };
}

/**
 * Format a picolorenzo count as Gnostic time string.
 * Example: "58a 1p 3k 2s 1b" = 58 aeons, 1 pleroma, 3 kenomas, 2 sophias, 1 barbelo
 */
export function formatGnosticTime(pLo: number): string {
  const d = decomposeGnosticTime(pLo);
  const parts: string[] = [];
  if (d.aeons > 0) parts.push(`${d.aeons}a`);
  if (d.pleromas > 0) parts.push(`${d.pleromas}p`);
  if (d.kenomas > 0) parts.push(`${d.kenomas}k`);
  if (d.sophias > 0) parts.push(`${d.sophias}s`);
  if (d.barbelos > 0) parts.push(`${d.barbelos}b`);
  return parts.join(' ') || '0b';
}

/**
 * Format with both Gnostic and ISO for human readability.
 * Example: "58a 1p 3k 2s 1b (2026-03-22T05:00:00.000Z)"
 */
export function formatDual(pLo: number): string {
  const date = fromPicolorenzos(pLo);
  return `${formatGnosticTime(pLo)} (${date.toISOString()})`;
}

// ── Duration helpers ─────────────────────────────────────────────────────────

/** Convert a duration in seconds to picolorenzos. */
export function secondsToPicolorenzos(seconds: number): number {
  return seconds / PICOLORENZO_SECONDS;
}

/** Convert a duration in picolorenzos to seconds. */
export function picolorenzosToSeconds(pLo: number): number {
  return pLo * PICOLORENZO_SECONDS;
}

/** Duration between two dates in picolorenzos. */
export function durationPicolorenzos(start: Date, end: Date): number {
  return (end.getTime() - start.getTime()) / PICOLORENZO_MS;
}

/** Duration between two ISO strings in picolorenzos. */
export function durationFromIso(startIso: string, endIso: string): number {
  return durationPicolorenzos(new Date(startIso), new Date(endIso));
}

// ── Solar Geometry (core) ───────────────────────────────────────────────────
//
// The sun's position is deterministic from time and observer location.
// These functions are part of the Gnostic Time System because any
// picolorenzo timestamp can answer "is it daylight?" without external
// dependencies. The sun does not change position -- only time does.

const _DEG_TO_RAD = Math.PI / 180;
const _RAD_TO_DEG = 180 / Math.PI;

/** Earth's axial tilt (obliquity of the ecliptic), degrees. J2000. */
export const OBLIQUITY_DEG = 23.4393;

/** Twilight zone classification by solar altitude angle. */
export type TwilightZone = 'day' | 'civil' | 'nautical' | 'astronomical' | 'night';

/** Observer on Earth's surface. */
export interface SolarObserver {
  readonly latitudeDeg: number;   // -90 to 90
  readonly longitudeDeg: number;  // -180 to 180
}

/** Solar position result. */
export interface SolarPosition {
  readonly altitudeDeg: number;
  readonly azimuthDeg: number;
  readonly zone: TwilightZone;
  readonly zoneFraction: number;  // 0-1 within current zone
}

/**
 * Approximate solar declination from day-of-year.
 * δ = -23.4393 × cos(360/365 × (doy + 10))
 */
export function solarDeclination(dayOfYear: number): number {
  return -OBLIQUITY_DEG * Math.cos(_DEG_TO_RAD * (360 / 365) * (dayOfYear + 10));
}

/**
 * Equation of time in minutes (orbital eccentricity + axial tilt correction).
 */
export function equationOfTime(dayOfYear: number): number {
  const b = _DEG_TO_RAD * (360 / 365) * (dayOfYear - 81);
  return 9.87 * Math.sin(2 * b) - 7.53 * Math.cos(b) - 1.5 * Math.sin(b);
}

/**
 * Classify twilight zone from solar altitude angle.
 *
 * | Altitude   | Zone               |
 * |------------|---------------------|
 * | > 0        | day                 |
 * | 0 to -6    | civil twilight      |
 * | -6 to -12  | nautical twilight   |
 * | -12 to -18 | astronomical        |
 * | < -18      | night               |
 */
export function classifyTwilightZone(altitudeDeg: number): TwilightZone {
  if (altitudeDeg > 0) return 'day';
  if (altitudeDeg > -6) return 'civil';
  if (altitudeDeg > -12) return 'nautical';
  if (altitudeDeg > -18) return 'astronomical';
  return 'night';
}

/** Zone altitude boundaries. */
const ZONE_BOUNDS: Record<TwilightZone, { min: number; max: number }> = {
  day:           { min: 0, max: 90 },
  civil:         { min: -6, max: 0 },
  nautical:      { min: -12, max: -6 },
  astronomical:  { min: -18, max: -12 },
  night:         { min: -90, max: -18 },
};

/**
 * Normalized 0-1 position within the current twilight zone.
 * 0 = lower boundary, 1 = upper boundary.
 */
export function twilightZoneFraction(altitudeDeg: number): number {
  const zone = classifyTwilightZone(altitudeDeg);
  const bounds = ZONE_BOUNDS[zone];
  const range = bounds.max - bounds.min;
  if (range === 0) return 1;
  return Math.max(0, Math.min(1, (altitudeDeg - bounds.min) / range));
}

/** Day-of-year (1-based) from epoch milliseconds. */
function _dayOfYear(epochMs: number): number {
  const date = new Date(epochMs);
  const startOfYear = Date.UTC(date.getUTCFullYear(), 0, 1);
  return Math.floor((epochMs - startOfYear) / 86_400_000) + 1;
}

/** Normalize degrees to [0, 360). */
function _normDeg(deg: number): number {
  return ((deg % 360) + 360) % 360;
}

/**
 * Compute solar position (altitude + azimuth) at observer location and time.
 *
 * Uses the analytic day-of-year approximation for solar declination.
 * Accurate to ~1 degree -- sufficient for daylight/twilight classification.
 */
export function solarPosition(observer: SolarObserver, epochMs: number): SolarPosition {
  const doy = _dayOfYear(epochMs);
  const decDeg = solarDeclination(doy);
  const eot = equationOfTime(doy);

  const date = new Date(epochMs);
  const utcHours =
    date.getUTCHours() +
    date.getUTCMinutes() / 60 +
    date.getUTCSeconds() / 3600;

  // Solar time: each degree of longitude = 4 minutes
  const solarTimeMinutes = utcHours * 60 + observer.longitudeDeg * 4 + eot;
  const hourAngleDeg = solarTimeMinutes / 4 - 180;

  const latRad = observer.latitudeDeg * _DEG_TO_RAD;
  const decRad = decDeg * _DEG_TO_RAD;
  const haRad = hourAngleDeg * _DEG_TO_RAD;

  // sin(alt) = sin(lat)*sin(dec) + cos(lat)*cos(dec)*cos(ha)
  const sinAlt =
    Math.sin(latRad) * Math.sin(decRad) +
    Math.cos(latRad) * Math.cos(decRad) * Math.cos(haRad);
  const altitudeDeg = Math.asin(Math.max(-1, Math.min(1, sinAlt))) * _RAD_TO_DEG;

  // Azimuth from north, clockwise
  const cosAlt = Math.cos(altitudeDeg * _DEG_TO_RAD);
  const cosAz = cosAlt > 0.001
    ? (Math.sin(decRad) - Math.sin(latRad) * sinAlt) / (Math.cos(latRad) * cosAlt)
    : 0;
  let azimuthDeg = Math.acos(Math.max(-1, Math.min(1, cosAz))) * _RAD_TO_DEG;
  if (Math.sin(haRad) > 0) azimuthDeg = 360 - azimuthDeg;

  const zone = classifyTwilightZone(altitudeDeg);
  return {
    altitudeDeg,
    azimuthDeg: _normDeg(azimuthDeg),
    zone,
    zoneFraction: twilightZoneFraction(altitudeDeg),
  };
}

/**
 * Is it daylight at the given picolorenzo time and observer?
 * Core Gnostic Time function: sun above horizon = daylight.
 */
export function isDaylightAtPLo(pLo: number, observer: SolarObserver): boolean {
  const epochMs = pLo * PICOLORENZO_MS;
  return solarPosition(observer, epochMs).zone === 'day';
}

/**
 * Full solar position at a picolorenzo timestamp.
 */
export function solarPositionAtPLo(pLo: number, observer: SolarObserver): SolarPosition {
  return solarPosition(observer, pLo * PICOLORENZO_MS);
}

// ── Solar Phase (0/1 binary fold) ───────────────────────────────────────────
//
// Daylight collapses to a single bit: 0 = night, 1 = day.
// This allows any time window to be folded into two halves --
// the lit phase and the dark phase -- like AM/PM but determined
// by the sun, not the clock.

/** Solar phase: 1 = daylight, 0 = night (including all twilight). */
export type SolarPhase = 0 | 1;

/**
 * Binary daylight phase at a picolorenzo timestamp.
 * 1 = sun above horizon, 0 = sun below horizon.
 */
export function solarPhase(pLo: number, observer: SolarObserver): SolarPhase {
  return isDaylightAtPLo(pLo, observer) ? 1 : 0;
}

/**
 * Solar phase including twilight gradation.
 * Returns a continuous value in [0, 1]:
 *   0 = deep night (altitude < -18)
 *   0..0.5 = twilight zones (ascending through astronomical, nautical, civil)
 *   0.5 = horizon (altitude = 0)
 *   0.5..1 = daytime (altitude 0 to 90)
 *
 * This maps the full -90..+90 degree altitude range to [0, 1],
 * providing a smooth gradient rather than a hard binary cutoff.
 */
export function solarPhaseContinuous(pLo: number, observer: SolarObserver): number {
  const pos = solarPositionAtPLo(pLo, observer);
  // Map altitude from [-90, 90] to [0, 1]
  return Math.max(0, Math.min(1, (pos.altitudeDeg + 90) / 180));
}

/** Solar half-day decomposition. Folds a time window into lit and dark halves. */
export interface SolarHalfDay {
  /** The picolorenzo midpoint of the lit phase (solar noon). */
  readonly solarNoonPLo: number;
  /** The picolorenzo midpoint of the dark phase (solar midnight). */
  readonly solarMidnightPLo: number;
  /** Duration of the lit phase in picolorenzos. */
  readonly litDurationPLo: number;
  /** Duration of the dark phase in picolorenzos. */
  readonly darkDurationPLo: number;
  /** Ratio of lit to total (0-1). At equator equinox this is ~0.5. */
  readonly litFraction: number;
}

/**
 * Fold a one-day window (starting at the given pLo) into its lit and dark halves.
 *
 * Samples the solar phase across the day at minute resolution and finds
 * the transitions. The result gives solar noon/midnight positions and
 * the duration of each half.
 *
 * One solar day ≈ 1/π picolorenzos (since 1 pLo = π days).
 */
export function solarHalfDay(startPLo: number, observer: SolarObserver): SolarHalfDay {
  const oneSolarDayPLo = 1 / Math.PI; // ~0.3183 pLo = 1 day
  const samples = 144; // every 10 minutes
  const stepPLo = oneSolarDayPLo / samples;

  let litCount = 0;
  let maxAlt = -Infinity;
  let minAlt = Infinity;
  let noonPLo = startPLo;
  let midnightPLo = startPLo;

  for (let i = 0; i < samples; i++) {
    const pLo = startPLo + i * stepPLo;
    const pos = solarPositionAtPLo(pLo, observer);
    if (pos.zone === 'day') litCount++;
    if (pos.altitudeDeg > maxAlt) {
      maxAlt = pos.altitudeDeg;
      noonPLo = pLo;
    }
    if (pos.altitudeDeg < minAlt) {
      minAlt = pos.altitudeDeg;
      midnightPLo = pLo;
    }
  }

  const litFraction = litCount / samples;
  return {
    solarNoonPLo: noonPLo,
    solarMidnightPLo: midnightPLo,
    litDurationPLo: litFraction * oneSolarDayPLo,
    darkDurationPLo: (1 - litFraction) * oneSolarDayPLo,
    litFraction,
  };
}
