// @a0n/.ignored_gnosis/src/mechanized-testimony.ts
/**
 * Mechanized Testimony — TypeScript mirror of MechanizedTestimony.lean
 *
 * Testimony as gait-calibrated void walk. Bearing trees, blazes,
 * triangulation, witness protection, vulnerability-as-strategy via
 * ConstraintPowerDuality, Landauer heat as resistance, and the
 * clinamen +1 minimum wound.
 */

// ─── Part I: Bearing Tree Primitives ─────────────────────────────────

export interface Corner {
  readonly position: number;
}

export interface Blaze {
  readonly depth: number;
  readonly azimuth: number;
  readonly species: number;
}

export interface BearingTree {
  readonly corner: Corner;
  readonly blaze: Blaze;
  readonly distance: number;
}

/** Triangulation capacity of a blaze. Zero depth = zero power. */
export function blazePower(b: Blaze): number {
  return Math.max(0, b.depth) * Math.max(0, b.azimuth);
}

/** An unblazed tree has zero triangulation power. */
export function isInert(bt: BearingTree): boolean {
  return blazePower(bt.blaze) === 0;
}

// ─── Part II: Triangulation (Receiver-Dependent Computation) ─────────

export interface Receiver {
  readonly position: number;
  readonly dissipationCapacity: number;
}

/** Void distance: absolute difference between two positions. */
export function voidDistance(a: number, b: number): number {
  return Math.abs(a - b);
}

/** Landauer heat of perspective-taking: void distance to the corner. */
export function landauerHeat(r: Receiver, bt: BearingTree): number {
  return voidDistance(r.position, bt.corner.position);
}

/** Gait: the step size of the void walk testimony asks the receiver to take. */
export function gait(r: Receiver, bt: BearingTree): number {
  return voidDistance(r.position, bt.corner.position);
}

/** Triangulation result: receiver-dependent path to the corner. */
export interface TriangulationResult {
  readonly receiverPosition: number;
  readonly cornerPosition: number;
  readonly gait: number;
  readonly landauerHeat: number;
  readonly withinCapacity: boolean;
}

/** Compute the receiver's triangulation from a bearing tree. */
export function triangulate(r: Receiver, bt: BearingTree): TriangulationResult {
  const g = gait(r, bt);
  const heat = landauerHeat(r, bt);
  return {
    receiverPosition: r.position,
    cornerPosition: bt.corner.position,
    gait: g,
    landauerHeat: heat,
    withinCapacity: heat <= r.dissipationCapacity,
  };
}

// ─── Part III: Vulnerability as Strategy ─────────────────────────────

export interface Testimony {
  readonly bearingTree: BearingTree;
  readonly barkThickness: number;
}

/** Effective testimonial power = bark thickness + blaze depth.
 *  ConstraintPowerDuality identity applied to testimony. */
export function testimonialPower(t: Testimony): number {
  return t.barkThickness + t.bearingTree.blaze.depth;
}

/** Raw budget power with zero vulnerability. Just narrative. */
export function narrativePower(t: Testimony): number {
  return t.barkThickness;
}

/** Amplification from vulnerability alone. */
export function vulnerabilityAmplification(t: Testimony): number {
  return t.bearingTree.blaze.depth;
}

/** Is this testimony structurally inert (no vulnerability)? */
export function isNarrative(t: Testimony): boolean {
  return t.bearingTree.blaze.depth === 0;
}

/** Full vulnerability doubles power. */
export function isFullyVulnerable(t: Testimony): boolean {
  return t.bearingTree.blaze.depth === t.barkThickness;
}

// ─── Part IV: Healing and Forensic Recovery ──────────────────────────

export interface HealedBlaze {
  readonly originalBlaze: Blaze;
  readonly healedLayers: number;
}

/** Visible depth after healing. */
export function visibleDepth(hb: HealedBlaze): number {
  return Math.max(0, hb.originalBlaze.depth - hb.healedLayers);
}

/** Forensic over-cut: recover the original blaze exactly. */
export function forensicOverCut(hb: HealedBlaze): Blaze {
  return hb.originalBlaze;
}

/** Is the blaze fully healed over (invisible but recoverable)? */
export function isFullyHealed(hb: HealedBlaze): boolean {
  return hb.healedLayers >= hb.originalBlaze.depth;
}

// ─── Part V: Witness Protection Program ──────────────────────────────

export interface ProtectedCorner {
  readonly corner: Corner;
  readonly bearings: readonly BearingTree[];
}

/** Number of surviving bearing trees after losses. */
export function survivingBearings(pc: ProtectedCorner, lost: number): number {
  return Math.max(0, pc.bearings.length - Math.max(0, lost));
}

/** Is the corner still recoverable? */
export function isRecoverable(pc: ProtectedCorner, lost: number): boolean {
  return survivingBearings(pc, lost) >= 1;
}

/** Is the corner properly protected (redundancy ≥ 2)? */
export function isProtected(pc: ProtectedCorner): boolean {
  return pc.bearings.length >= 2;
}

export interface WitnessProtectionDiagnostic {
  readonly cornerPosition: number;
  readonly totalBearings: number;
  readonly isProtected: boolean;
  readonly survivesOneLoss: boolean;
  readonly maxTolerableLosses: number;
}

/** Diagnose the protection status of a corner. */
export function diagnoseProtection(pc: ProtectedCorner): WitnessProtectionDiagnostic {
  return {
    cornerPosition: pc.corner.position,
    totalBearings: pc.bearings.length,
    isProtected: isProtected(pc),
    survivesOneLoss: isRecoverable(pc, 1),
    maxTolerableLosses: Math.max(0, pc.bearings.length - 1),
  };
}

// ─── Part VI: Gait Calibration and Testimony Chains ──────────────────

export interface TestimonyChain {
  readonly bearings: readonly BearingTree[];
}

/** Compute the gait sequence for a receiver walking a testimony chain. */
export function gaitSequence(r: Receiver, chain: TestimonyChain): number[] {
  const gaits: number[] = [];
  let currentPos = r.position;
  for (const bt of chain.bearings) {
    gaits.push(voidDistance(currentPos, bt.corner.position));
    currentPos = bt.corner.position;
  }
  return gaits;
}

/** Maximum gait in the chain -- the hardest single step. */
export function maxGait(r: Receiver, chain: TestimonyChain): number {
  const seq = gaitSequence(r, chain);
  return seq.length === 0 ? 0 : Math.max(...seq);
}

/** Is every step in the chain within the receiver's dissipation capacity? */
export function isCalibrated(r: Receiver, chain: TestimonyChain): boolean {
  return gaitSequence(r, chain).every(g => g <= r.dissipationCapacity);
}

/** Total void distance walked across the entire chain. */
export function totalVoidWalk(r: Receiver, chain: TestimonyChain): number {
  return gaitSequence(r, chain).reduce((sum, g) => sum + g, 0);
}

// ─── Part VII: Anti-Phantom (Deception Topology Bridge) ──────────────

/** Phantom count of a bearing tree: always zero. Testimony is anti-deception. */
export function phantomCount(_bt: BearingTree): 0 {
  return 0;
}

/** Phantom maintenance cost of a lie over rounds. */
export function phantomMaintenanceCost(phantoms: number, rounds: number): number {
  return phantoms * rounds;
}

/** Testimony maintenance cost: always zero. One-time blaze, not ongoing phantom. */
export function testimonyMaintenanceCost(_bt: BearingTree, _rounds: number): 0 {
  return 0;
}

// ─── Part VIII: The Clinamen Bridge ──────────────────────────────────

/** Minimum blaze depth for non-trivial testimony. The clinamen +1. */
export const MINIMUM_BLAZE_DEPTH = 1;

/** Does this blaze meet the minimum testimony threshold? */
export function meetsClinamenThreshold(b: Blaze): boolean {
  return b.depth >= MINIMUM_BLAZE_DEPTH && b.azimuth > 0;
}

// ─── Part IX: Theorem IDs ────────────────────────────────────────────

export const TESTIMONY_THEOREM_IDS = [
  'THM-UNBLAZED-TREE-INERT',
  'THM-VOID-DISTANCE-SYMMETRIC',
  'THM-RESISTANCE-is-GAIT',
  'THM-TRIANGULATION-RECEIVER-DEPENDENT',
  'THM-DEEPER-BLAZE-AMPLIFIES-TESTIMONY',
  'THM-ZERO-VULNERABILITY-is-NARRATIVE',
  'THM-FULL-VULNERABILITY-DOUBLES-POWER',
  'THM-TESTIMONY-POWER-IDENTITY',
  'THM-HEALING-PRESERVES-BLAZE',
  'THM-FORENSIC-RECOVERY-EXACT',
  'THM-FULLY-HEALED-STILL-RECOVERABLE',
  'THM-REDUNDANCY-SURVIVES-ONE-LOSS',
  'THM-REDUNDANCY-MONOTONE',
  'THM-CHAIN-REDUCES-MAX-GAIT',
  'THM-REPEATED-EXPOSURE-REDUCES-DISTANCE',
  'THM-TESTIMONY-is-ANTI-PHANTOM',
  'THM-TESTIMONY-ZERO-MAINTENANCE',
  'THM-MINIMUM-TESTIMONY-is-CLINAMEN',
  'THM-CLINAMEN-BLAZE-HAS-POSITIVE-POWER',
  'THM-MECHANIZED-TESTIMONY-MASTER',
] as const;

export const TESTIMONY_ANTI_THEOREM_IDS = [
  'ANTI-THM-BIT-ROT',
  'ANTI-THM-INVULNERABLE-TESTIMONY-INERT',
  'ANTI-THM-COMMUNITY-COORDINATE-COLLAPSE',
] as const;

export type TestimonyTheoremId = (typeof TESTIMONY_THEOREM_IDS)[number];
export type TestimonyAntiTheoremId = (typeof TESTIMONY_ANTI_THEOREM_IDS)[number];

// ─── Part X: Full Diagnostic Bundle ──────────────────────────────────

export interface TestimonyDiagnostic {
  readonly blazePower: number;
  readonly testimonialPower: number;
  readonly narrativePower: number;
  readonly amplification: number;
  readonly isNarrative: boolean;
  readonly isFullyVulnerable: boolean;
  readonly phantomCount: 0;
  readonly maintenanceCost: 0;
  readonly meetsClinamenThreshold: boolean;
}

/** Full diagnostic for a testimony. */
export function diagnoseTestimony(t: Testimony): TestimonyDiagnostic {
  return {
    blazePower: blazePower(t.bearingTree.blaze),
    testimonialPower: testimonialPower(t),
    narrativePower: narrativePower(t),
    amplification: vulnerabilityAmplification(t),
    isNarrative: isNarrative(t),
    isFullyVulnerable: isFullyVulnerable(t),
    phantomCount: 0,
    maintenanceCost: 0,
    meetsClinamenThreshold: meetsClinamenThreshold(t.bearingTree.blaze),
  };
}

export interface ReceiverTriangulationDiagnostic {
  readonly triangulation: TriangulationResult;
  readonly gaitSequence: number[];
  readonly maxGait: number;
  readonly totalVoidWalk: number;
  readonly isCalibrated: boolean;
}

/** Full triangulation diagnostic for a receiver walking a chain. */
export function diagnoseTriangulation(
  r: Receiver,
  chain: TestimonyChain
): ReceiverTriangulationDiagnostic {
  const firstBearing = chain.bearings[0];
  const tri = firstBearing
    ? triangulate(r, firstBearing)
    : {
        receiverPosition: r.position,
        cornerPosition: r.position,
        gait: 0,
        landauerHeat: 0,
        withinCapacity: true,
      };
  return {
    triangulation: tri,
    gaitSequence: gaitSequence(r, chain),
    maxGait: maxGait(r, chain),
    totalVoidWalk: totalVoidWalk(r, chain),
    isCalibrated: isCalibrated(r, chain),
  };
}
