// @a0n/.ignored_gnosis/src/mesh-policy-edge-compiled.ts
/**
 * Mesh-policy edge runtime — compiled-bundle flavor.
 *
 * Thin adapter that deserializes a `.bw` mesh-policy bundle emitted by
 * `./mesh-policy-compile.ts` and forwards to the raw-config
 * {@link wrapWithMeshPolicy} in `./mesh-policy-edge.ts`. Preserves the
 * identical `WrappedFetch` contract — downstream call sites are
 * interchangeable between the raw and compiled constructors.
 *
 * No Node-specific imports, no global-log calls, no Node process
 * globals. Version
 * mismatches surface via the optional `audit` callback when
 * `strictVersion === false`; otherwise they throw immediately.
 */

import {
  wrapWithMeshPolicy,
  type InnerFetch,
  type MeshPolicyConfig,
  type WrapOptions,
  type WrappedFetch,
} from './mesh-policy-edge';
import {
  inspectCompiledBundle as inspectBundleImpl,
  loadMeshPolicyFromBundle,
  MESH_POLICY_MANIFEST_FORMAT,
  type MeshPolicyCompiledManifest,
} from './mesh-policy-compile';

export type {
  MeshPolicyCompiledManifest,
} from './mesh-policy-compile';

/** Re-exported from `./mesh-policy-compile.js` so deploy tooling can
 *  import both the bundle runtime and the content-addressed fingerprint
 *  helper from the same entry point. */
export { fingerprintMeshPolicyBundle } from './mesh-policy-compile';

/** Deny-all fallback policy used when `strictVersion: false` encounters
 *  an unknown bundle format. Empty ports → every path 404s via
 *  `emptyPortListSealsRegion`. */
const DENY_ALL_POLICY: MeshPolicyConfig = { ports: [] };

export interface CompiledWrapOptions extends WrapOptions {
  /**
   * When `true` (default), bundles with an unrecognized
   * `manifest.format` cause {@link wrapWithMeshPolicyCompiled} to
   * throw. When `false`, the wrapper returns a deny-all handler and
   * emits an `audit` callback describing the fallback — useful during
   * rolling schema upgrades where a stale bundle is preferable to an
   * outage.
   */
  readonly strictVersion?: boolean;
}

/**
 * Inspect a `.bw` bundle without committing to a full deserialize.
 * Re-exports the compile-side helper so consumers can import a single
 * entry point.
 */
export function inspectCompiledBundle(
  bwBytes: Uint8Array,
): MeshPolicyCompiledManifest {
  return inspectBundleImpl(bwBytes);
}

/**
 * Wrap an inner Web Fetch handler with mesh-policy enforcement from a
 * compiled `.bw` bundle.
 *
 * Contract: `wrapWithMeshPolicyCompiled(inner, compileMeshPolicy(p).bwBytes, opts)`
 * behaves identically to `wrapWithMeshPolicy(inner, p, opts)` for all
 * policies `p` and all request shapes. Round-trip parity is the
 * entire point of the compiled pipeline.
 *
 * @throws `Error` when the bundle's manifest format is unknown and
 *         `options.strictVersion !== false`.
 */
export function wrapWithMeshPolicyCompiled(
  innerFetch: InnerFetch,
  bwBytes: Uint8Array,
  options?: CompiledWrapOptions,
): WrappedFetch {
  const { policy, manifest } = loadMeshPolicyFromBundle(bwBytes);

  const strictVersion = options?.strictVersion !== false;
  if (manifest.format !== MESH_POLICY_MANIFEST_FORMAT) {
    if (strictVersion) {
      throw new Error(
        `mesh-policy-edge-compiled: unknown manifest.format '${manifest.format}' ` +
          `(expected '${MESH_POLICY_MANIFEST_FORMAT}')`,
      );
    }
    if (options?.audit) {
      options.audit(
        `[mesh-policy-edge-compiled] unknown-format ` +
          `got='${manifest.format}' expected='${MESH_POLICY_MANIFEST_FORMAT}' fallback=deny-all`,
      );
    }
    return wrapWithMeshPolicy(innerFetch, DENY_ALL_POLICY, options);
  }

  return wrapWithMeshPolicy(innerFetch, policy, options);
}
