// @a0n/.ignored_gnosis/src/module-system-handlers.ts
/**
 * Module System GG Topology Handlers
 *
 * Handlers for module manifest creation, dependency resolution,
 * lockfile management, and compatibility checking.
 */

import type { GnosisRegistry } from './runtime/registry';
import {
  parseSemVer,
  formatSemVer,
  parseConstraint,
  satisfiesConstraint,
  createManifest,
  createLockfile,
  addLockfileEntry,
  resolveDependencies,
  checkModuleCompatibility,
  type ModuleManifest,
  type SemVer,
  type Lockfile,
} from './module-system';

/**
 * Handles the gnosis register Module System Handlers workflow.
 */
export function registerModuleSystemHandlers(registry: GnosisRegistry): void {
  let activeLockfile = createLockfile();

  // Parse a semver string
  registry.register('SemVerParser', async (_payload, props) => {
    const version = props.version ?? '';
    const parsed = parseSemVer(version);
    return parsed
      ? { ...parsed, formatted: formatSemVer(parsed) }
      : { error: `Invalid version: ${version}` };
  });

  // Check version constraint satisfaction
  registry.register('VersionConstraintChecker', async (_payload, props) => {
    const version = parseSemVer(props.version ?? '');
    const constraint = parseConstraint(props.constraint ?? '*');
    if (!version || !constraint) {
      return { error: 'Invalid version or constraint' };
    }
    return {
      version: formatSemVer(version),
      constraint: props.constraint,
      satisfies: satisfiesConstraint(version, constraint),
    };
  });

  // Create a module manifest
  registry.register('ManifestCreator', async (_payload, props) => {
    const name = props.name ?? '@gnosis/unnamed';
    const version = props.version ?? '0.1.0';
    try {
      const manifest = createManifest(name, version);

      // Parse dependencies from props
      const depsRaw = props.dependencies ?? '';
      if (depsRaw) {
        for (const dep of depsRaw.split(';')) {
          const [depName, constraintStr] = dep.split('@').map((s) => s.trim());
          if (depName && constraintStr) {
            const constraint = parseConstraint(constraintStr);
            if (constraint) {
              manifest.dependencies.push({
                name: depName,
                constraint,
                dev: false,
              });
            }
          }
        }
      }

      // Parse exports
      const exportsRaw = props.exports ?? '';
      if (exportsRaw) {
        manifest.exports = exportsRaw
          .split(',')
          .map((s) => s.trim())
          .filter(Boolean);
      }

      return manifest;
    } catch (e: any) {
      return { error: e.message };
    }
  });

  // Resolve dependencies
  registry.register('DependencyResolver', async (payload, props) => {
    const manifest = payload as ModuleManifest;
    if (!manifest?.name) {
      return { error: 'No manifest provided' };
    }

    // Build available versions from props
    const available = new Map<string, SemVer[]>();
    const availableRaw = props.available ?? '';
    for (const entry of availableRaw.split(';')) {
      const [name, versions] = entry.split('=').map((s) => s.trim());
      if (name && versions) {
        const parsed = versions
          .split(',')
          .map((v) => parseSemVer(v.trim()))
          .filter((v): v is SemVer => v !== null);
        available.set(name, parsed);
      }
    }

    return resolveDependencies(manifest, available);
  });

  // Initialize lockfile
  registry.register('LockfileInit', async () => {
    activeLockfile = createLockfile();
    return { initialized: true, version: activeLockfile.lockfileVersion };
  });

  // Add entry to lockfile
  registry.register('LockfileAdd', async (payload) => {
    const entry = payload as {
      name: string;
      version: string;
      integrity: string;
      dependencies: Record<string, string>;
    };
    const version = parseSemVer(entry.version);
    if (!version) return { error: `Invalid version: ${entry.version}` };
    addLockfileEntry(activeLockfile, {
      name: entry.name,
      version,
      integrity: entry.integrity,
      dependencies: entry.dependencies ?? {},
    });
    return { added: true, key: `${entry.name}@${entry.version}` };
  });

  // Check module compatibility
  registry.register('ModuleCompatibilityChecker', async (payload) => {
    const { consumer, provider } = payload as {
      consumer: ModuleManifest;
      provider: ModuleManifest;
    };
    return checkModuleCompatibility(consumer, provider);
  });

  // Get lockfile status
  registry.register('LockfileStatus', async () => {
    return {
      version: activeLockfile.lockfileVersion,
      entries: activeLockfile.entries.size,
      packages: [...activeLockfile.entries.keys()],
    };
  });
}

export const MODULE_SYSTEM_FEATURES = [
  'semver-parsing',
  'constraint-checking',
  'manifest-creation',
  'dependency-resolution',
  'lockfile-management',
  'compatibility-checking',
] as const;
