// @a0n/.ignored_gnosis/src/native-host-bridge.ts
import { TextDecoder, TextEncoder } from 'node:util';
import { fileURLToPath } from 'node:url';
import { resolve } from 'node:path';
import {
  GnosisGnodeExecutionSurface,
  type GnodeExecutionControlRequest,
  type GnodeExecutionControlResponse,
} from './gnode-execution';
import {
  GnosisWasmExecutionSurface,
  type WasmExecutionControlRequest,
  type WasmExecutionControlResponse,
} from './wasm-execution';

interface NativeHostBridgeRequest {
  readonly method: string;
  readonly path: string;
  readonly query?: string;
  readonly bodyText?: string;
  readonly headers?: Record<string, string>;
}

interface NativeHostBridgeResponse {
  readonly status: number;
  readonly headers: Record<string, string>;
  readonly body: unknown;
}

const textDecoder = new TextDecoder('utf-8', { fatal: true });
const textEncoder = new TextEncoder();

function decodeJsonBody(body: Uint8Array): unknown {
  const text = textDecoder.decode(body);
  if (text.length === 0) {
    return null;
  }

  return JSON.parse(text) as unknown;
}

function encodeRequestBody(bodyText?: string): Uint8Array | undefined {
  if (bodyText === undefined) {
    return undefined;
  }

  return textEncoder.encode(bodyText);
}

function toBridgeResponse(
  response: GnodeExecutionControlResponse | WasmExecutionControlResponse
): NativeHostBridgeResponse {
  return {
    status: response.status,
    headers: response.headers,
    body: decodeJsonBody(response.body),
  };
}

function toControlRequest(
  request: NativeHostBridgeRequest
): GnodeExecutionControlRequest {
  return {
    method: request.method,
    path: request.path,
    ...(request.query ? { query: request.query } : {}),
    ...(request.headers ? { headers: request.headers } : {}),
    ...(request.bodyText !== undefined
      ? { body: encodeRequestBody(request.bodyText) }
      : {}),
  };
}

function isMainModule(): boolean {
  if (!process.argv[1]) {
    return false;
  }

  return resolve(fileURLToPath(import.meta.url)) === resolve(process.argv[1]);
}

/**
 * Runs the gnosis command-line workflow.
 */
export async function main(argv = process.argv.slice(2)): Promise<number> {
  const enableGnode = argv.length === 0 || argv.includes('--gnode');
  const enableWasm = argv.length === 0 || argv.includes('--wasm');
  const gnodeSurface = new GnosisGnodeExecutionSurface({
    enabled: enableGnode,
  });
  const wasmSurface = new GnosisWasmExecutionSurface({
    enabled: enableWasm,
  });

  process.stdin.setEncoding('utf8');

  let buffer = '';
  for await (const chunk of process.stdin) {
    buffer += chunk;
    while (true) {
      const newlineIndex = buffer.indexOf('\n');
      if (newlineIndex < 0) {
        break;
      }

      const line = buffer.slice(0, newlineIndex).trim();
      buffer = buffer.slice(newlineIndex + 1);
      if (line.length === 0) {
        continue;
      }

      const response = await handleLine(line, gnodeSurface, wasmSurface);
      process.stdout.write(`${JSON.stringify(response)}\n`);
    }
  }

  const trailing = buffer.trim();
  if (trailing.length > 0) {
    const response = await handleLine(trailing, gnodeSurface, wasmSurface);
    process.stdout.write(`${JSON.stringify(response)}\n`);
  }

  return 0;
}

async function handleLine(
  line: string,
  gnodeSurface: GnosisGnodeExecutionSurface,
  wasmSurface: GnosisWasmExecutionSurface
): Promise<NativeHostBridgeResponse | { error: string }> {
  try {
    const request = JSON.parse(line) as NativeHostBridgeRequest;
    const controlRequest = toControlRequest(request);
    const response = request.path.startsWith('/.aeon/gnode/')
      ? await gnodeSurface.handleRequest(controlRequest)
      : request.path.startsWith('/.aeon/wasm/')
      ? await wasmSurface.handleRequest(
          controlRequest as WasmExecutionControlRequest
        )
      : null;

    if (!response) {
      return { error: `Unsupported native host path '${request.path}'.` };
    }

    return toBridgeResponse(response);
  } catch (error) {
    return {
      error: error instanceof Error ? error.message : String(error),
    };
  }
}

if (isMainModule()) {
  main().then(
    (code) => {
      process.exitCode = code;
    },
    (error: unknown) => {
      process.stderr.write(
        `${
          error instanceof Error ? error.stack ?? error.message : String(error)
        }\n`
      );
      process.exitCode = 1;
    }
  );
}
