// @a0n/.ignored_gnosis/src/native-execution-auth.ts
import { createHash } from 'node:crypto';
import {
  mergeExecutionAuthContexts,
  normalizeExecutionAuthContext,
  type GnosisCapability,
  type GnosisExecutionAuthContext,
} from './auth/core';
import {
  createReactorHandleGrant,
  verifyReactorHandleGrant,
  type ReactorHandleGrantClaims,
  type ReactorHandleGrantEnvelope,
} from './auth/reactor-attestation';

export interface NativeExecutionControlHeaders {
  readonly [name: string]: string | undefined;
}

export interface ResolvedNativeExecutionAuth {
  readonly executionAuth: GnosisExecutionAuthContext | null;
  readonly source: 'none' | 'headers' | 'explicit' | 'merged';
}

export interface NativeExecutionAuthSummary {
  readonly enforced: boolean;
  readonly principal: string | null;
  readonly capabilityCount: number;
  readonly tokenPresent: boolean;
  readonly reactorHandleGrantCount: number;
  readonly reactorHandleKinds: readonly string[];
  readonly source: ResolvedNativeExecutionAuth['source'];
}

export interface NativeExecutionReactorGrantSigner {
  readonly keyId: string;
  readonly privateKey: JsonWebKey;
  readonly publicKey: JsonWebKey;
}

const REACTOR_GRANT_SIGNING_ALGORITHM: EcKeyGenParams = {
  name: 'ECDSA',
  namedCurve: 'P-256',
};
const DEFAULT_NATIVE_REACTOR_GRANT_AUDIENCE = 'aeon://gnosis/native-execution';
const seenForwardedReactorHandleGrantReplays = new Map<string, number>();

function normalizeHeaders(
  headers: NativeExecutionControlHeaders | undefined
): Record<string, string> {
  if (!headers) {
    return {};
  }

  const normalized: Record<string, string> = {};
  for (const [name, value] of Object.entries(headers)) {
    if (typeof value !== 'string') {
      continue;
    }

    const trimmedName = name.trim().toLowerCase();
    const trimmedValue = value.trim();
    if (!trimmedName || !trimmedValue) {
      continue;
    }

    normalized[trimmedName] = trimmedValue;
  }

  return normalized;
}

function parseBooleanHeader(value: string | undefined): boolean {
  if (!value) {
    return false;
  }

  const normalized = value.trim().toLowerCase();
  return normalized === '1' || normalized === 'true' || normalized === 'yes';
}

function parseCapabilityHeaderValue(value: string): GnosisCapability[] {
  if (!value.trim()) {
    return [];
  }

  if (value.trim().startsWith('[')) {
    try {
      const parsed = JSON.parse(value) as unknown;
      if (Array.isArray(parsed)) {
        return parsed
          .map((entry) => {
            if (typeof entry === 'string') {
              const capability = entry.trim();
              return capability
                ? {
                    can: capability,
                    with: '*',
                  }
                : null;
            }

            if (
              typeof entry === 'object' &&
              entry !== null &&
              typeof (entry as { can?: unknown }).can === 'string' &&
              typeof (entry as { with?: unknown }).with === 'string'
            ) {
              return {
                can: (entry as { can: string }).can,
                with: (entry as { with: string }).with,
              };
            }

            return null;
          })
          .filter((entry): entry is GnosisCapability => entry !== null);
      }
    } catch {
      // Fall through to comma parsing.
    }
  }

  return value
    .split(',')
    .map((entry) => entry.trim())
    .filter((entry) => entry.length > 0)
    .map((entry) => ({
      can: entry,
      with: '*',
    }));
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null;
}

function isNonEmptyString(value: unknown): value is string {
  return typeof value === 'string' && value.trim().length > 0;
}

function normalizeGrantStringList(values: unknown): string[] {
  if (!Array.isArray(values)) {
    return [];
  }

  return Array.from(
    new Set(
      values
        .filter((value): value is string => isNonEmptyString(value))
        .map((value) => value.trim())
    )
  ).sort((left, right) => left.localeCompare(right));
}

function parseReactorHandleGrantClaims(
  value: unknown
): ReactorHandleGrantClaims | null {
  if (!isRecord(value)) {
    return null;
  }

  if (
    !isNonEmptyString(value.handleId) ||
    !isNonEmptyString(value.kind) ||
    !isNonEmptyString(value.ownerDid) ||
    !isNonEmptyString(value.audience) ||
    !isNonEmptyString(value.resource) ||
    !isNonEmptyString(value.runtimeBindingHash) ||
    !Number.isFinite(value.issuedAt) ||
    !Number.isFinite(value.expiresAt)
  ) {
    return null;
  }

  return {
    handleId: value.handleId.trim(),
    kind: value.kind.trim() as ReactorHandleGrantClaims['kind'],
    ownerDid: value.ownerDid.trim(),
    audience: value.audience.trim(),
    resource: value.resource.trim(),
    operations: normalizeGrantStringList(
      value.operations
    ) as ReactorHandleGrantClaims['operations'],
    ucanCapabilities: normalizeGrantStringList(value.ucanCapabilities),
    runtimeBindingHash: value.runtimeBindingHash.trim(),
    ...(isNonEmptyString(value.delegationChainHash)
      ? { delegationChainHash: value.delegationChainHash.trim() }
      : {}),
    issuedAt: value.issuedAt as number,
    expiresAt: value.expiresAt as number,
  };
}

function parseReactorHandleGrantHeaderValue(
  value: string,
  principal: string | undefined,
  runtimeBindingHash: string | undefined
): ReactorHandleGrantEnvelope[] {
  if (!value.trim()) {
    return [];
  }

  try {
    const parsed = JSON.parse(value) as unknown;
    if (!Array.isArray(parsed)) {
      return [];
    }

    return parsed
      .map((entry) => {
        if (
          !isRecord(entry) ||
          entry.algorithm !== 'ES256' ||
          !isNonEmptyString(entry.signature)
        ) {
          return null;
        }

        const claims = parseReactorHandleGrantClaims(entry.claims);
        if (!claims) {
          return null;
        }
        if (principal && claims.ownerDid !== principal.trim()) {
          return null;
        }
        if (
          runtimeBindingHash &&
          claims.runtimeBindingHash !== runtimeBindingHash.trim()
        ) {
          return null;
        }

        return {
          algorithm: 'ES256',
          signature: entry.signature.trim(),
          claims,
          ...(isNonEmptyString(entry.keyId)
            ? { keyId: entry.keyId.trim() }
            : {}),
          ...(isRecord(entry.publicKey)
            ? { publicKey: entry.publicKey as JsonWebKey }
            : {}),
        } satisfies ReactorHandleGrantEnvelope;
      })
      .filter((entry): entry is ReactorHandleGrantEnvelope => entry !== null);
  } catch {
    return [];
  }
}

function pruneExpiredForwardedReactorGrantReplays(nowMs: number): void {
  for (const [replayKey, expiresAt] of seenForwardedReactorHandleGrantReplays) {
    if (expiresAt <= nowMs) {
      seenForwardedReactorHandleGrantReplays.delete(replayKey);
    }
  }
}

function forwardedGrantReplayKey(grant: ReactorHandleGrantEnvelope): string {
  return `${grant.claims.handleId}:${grant.signature}`;
}

function isScopedGrantOperationSet(claims: ReactorHandleGrantClaims): boolean {
  const operations = new Set(claims.operations);
  if (claims.kind === 'request') {
    return (
      claims.resource.endsWith('#response') === false &&
      operations.has('request.metadata') &&
      operations.has('request.read') &&
      claims.operations.every((operation) => operation.startsWith('request.'))
    );
  }

  if (claims.kind === 'response') {
    return (
      claims.resource.endsWith('#response') &&
      operations.has('response.start') &&
      operations.has('response.write') &&
      operations.has('response.end') &&
      claims.operations.every((operation) => operation.startsWith('response.'))
    );
  }

  return false;
}

async function verifyForwardedReactorHandleGrants(input: {
  readonly grants: readonly ReactorHandleGrantEnvelope[];
  readonly principal?: string;
  readonly runtimeBindingHash?: string;
  readonly expectedAudience?: string;
}): Promise<ReactorHandleGrantEnvelope[]> {
  if (input.grants.length === 0) {
    return [];
  }

  const nowMs = Date.now();
  const verified: ReactorHandleGrantEnvelope[] = [];
  for (const grant of input.grants) {
    const verification = await verifyReactorHandleGrant(grant, {
      nowMs,
      allowInlinePublicKey: true,
      ...(input.principal?.trim()
        ? { expectedOwnerDid: input.principal.trim() }
        : {}),
      ...(input.runtimeBindingHash?.trim()
        ? {
            expectedRuntimeBindingHash: input.runtimeBindingHash.trim(),
          }
        : {}),
      ...(input.expectedAudience?.trim()
        ? { expectedAudience: input.expectedAudience.trim() }
        : {}),
    });
    if (!verification.valid || !verification.claims) {
      continue;
    }
    if (!isScopedGrantOperationSet(verification.claims)) {
      continue;
    }

    verified.push({
      ...grant,
      claims: verification.claims,
    });
  }

  if (verified.length === 0) {
    return [];
  }

  const verifiedOwners = new Set(
    verified.map((grant) => grant.claims.ownerDid)
  );
  const effectivePrincipal = input.principal?.trim()
    ? input.principal.trim()
    : verifiedOwners.size === 1
    ? verified[0]?.claims.ownerDid
    : undefined;
  const principalScoped =
    effectivePrincipal === undefined
      ? []
      : verified.filter(
          (grant) => grant.claims.ownerDid === effectivePrincipal
        );

  pruneExpiredForwardedReactorGrantReplays(nowMs);

  const admitted: ReactorHandleGrantEnvelope[] = [];
  for (const grant of principalScoped) {
    const replayKey = forwardedGrantReplayKey(grant);
    const replayExpiresAt =
      seenForwardedReactorHandleGrantReplays.get(replayKey) ?? 0;
    if (replayExpiresAt > nowMs) {
      continue;
    }

    seenForwardedReactorHandleGrantReplays.set(
      replayKey,
      grant.claims.expiresAt
    );
    admitted.push(grant);
  }

  return admitted;
}

function serializeCapability(capability: GnosisCapability): string {
  return `${capability.can} ${capability.with}`;
}

/**
 * Handles the gnosis derive Execution Auth From Headers workflow.
 */
export async function deriveExecutionAuthFromHeaders(
  headers: NativeExecutionControlHeaders | undefined,
  options: {
    readonly expectedReactorGrantAudience?: string;
  } = {}
): Promise<GnosisExecutionAuthContext | null> {
  const normalized = normalizeHeaders(headers);
  const capabilities = parseCapabilityHeaderValue(
    normalized['x-ucan-capabilities'] ?? ''
  );
  const runtimeBindingHash = normalized['x-aeon-runtime-binding-hash'];
  const headerPrincipal = normalized['x-aeon-actor'];
  const parsedReactorHandleGrants = parseReactorHandleGrantHeaderValue(
    normalized['x-aeon-reactor-handle-grants'] ?? '',
    headerPrincipal,
    runtimeBindingHash
  );
  const reactorGrantInput: {
    grants: readonly ReactorHandleGrantEnvelope[];
    principal?: string;
    runtimeBindingHash?: string;
    expectedAudience: string;
  } = {
    grants: parsedReactorHandleGrants,
    expectedAudience:
      options.expectedReactorGrantAudience ??
      DEFAULT_NATIVE_REACTOR_GRANT_AUDIENCE,
  };
  if (headerPrincipal !== undefined) {
    reactorGrantInput.principal = headerPrincipal;
  }
  if (runtimeBindingHash !== undefined) {
    reactorGrantInput.runtimeBindingHash = runtimeBindingHash;
  }
  const reactorHandleGrants = await verifyForwardedReactorHandleGrants(
    reactorGrantInput
  );
  const principal =
    headerPrincipal ??
    (() => {
      const owners = Array.from(
        new Set(reactorHandleGrants.map((grant) => grant.claims.ownerDid))
      );
      return owners.length === 1 ? owners[0] : undefined;
    })();
  const token = normalized.authorization;
  const enforce =
    parseBooleanHeader(normalized['x-aeon-auth-verified']) ||
    capabilities.length > 0 ||
    reactorHandleGrants.length > 0;

  if (
    !principal &&
    !token &&
    capabilities.length === 0 &&
    reactorHandleGrants.length === 0
  ) {
    return null;
  }

  return normalizeExecutionAuthContext({
    enforce,
    principal,
    token,
    capabilities,
    reactorHandleGrants,
  });
}

/**
 * Resolves the Native Execution Auth.
 */
export async function resolveNativeExecutionAuth(options: {
  readonly headers?: NativeExecutionControlHeaders;
  readonly explicitExecutionAuth?: GnosisExecutionAuthContext | null;
  readonly expectedReactorGrantAudience?: string;
}): Promise<ResolvedNativeExecutionAuth> {
  const headerAuth = await deriveExecutionAuthFromHeaders(
    options.headers,
    options.expectedReactorGrantAudience !== undefined
      ? {
          expectedReactorGrantAudience: options.expectedReactorGrantAudience,
        }
      : {}
  );
  const explicitAuth = normalizeExecutionAuthContext(
    options.explicitExecutionAuth
  );

  if (headerAuth && explicitAuth) {
    return {
      executionAuth: mergeExecutionAuthContexts(headerAuth, explicitAuth),
      source: 'merged',
    };
  }

  if (explicitAuth) {
    return {
      executionAuth: explicitAuth,
      source: 'explicit',
    };
  }

  if (headerAuth) {
    return {
      executionAuth: headerAuth,
      source: 'headers',
    };
  }

  return {
    executionAuth: null,
    source: 'none',
  };
}

/**
 * Handles the gnosis summarize Native Execution Auth workflow.
 */
export function summarizeNativeExecutionAuth(
  auth: GnosisExecutionAuthContext | null,
  source: ResolvedNativeExecutionAuth['source'] = 'none'
): NativeExecutionAuthSummary | undefined {
  if (!auth) {
    return undefined;
  }

  return {
    enforced: auth.enforce === true,
    principal: auth.principal ?? null,
    capabilityCount: auth.capabilities.length,
    tokenPresent: typeof auth.token === 'string' && auth.token.length > 0,
    reactorHandleGrantCount: auth.reactorHandleGrants?.length ?? 0,
    reactorHandleKinds: Array.from(
      new Set(
        (auth.reactorHandleGrants ?? []).map((grant) => grant.claims.kind)
      )
    ).sort((left, right) => left.localeCompare(right)),
    source,
  };
}

/**
 * Handles the gnosis generate Native Execution Reactor Grant Signer workflow.
 */
export async function generateNativeExecutionReactorGrantSigner(
  keyIdSeed = 'gnosis-native-execution'
): Promise<NativeExecutionReactorGrantSigner> {
  const keyPair = await crypto.subtle.generateKey(
    REACTOR_GRANT_SIGNING_ALGORITHM,
    true,
    ['sign', 'verify']
  );
  const publicKey = await crypto.subtle.exportKey('jwk', keyPair.publicKey);
  const privateKey = await crypto.subtle.exportKey('jwk', keyPair.privateKey);
  const keyId = createHash('sha256')
    .update(`${keyIdSeed}:${JSON.stringify(publicKey)}`)
    .digest('hex')
    .slice(0, 16);

  return {
    keyId,
    privateKey,
    publicKey,
  };
}

/**
 * Handles the gnosis attach Request Scoped Reactor Handle Grants workflow.
 */
export async function attachRequestScopedReactorHandleGrants(input: {
  readonly executionAuth: GnosisExecutionAuthContext | null;
  readonly runtimeBindingHash?: string | null;
  readonly audience: string;
  readonly method: string;
  readonly path: string;
  readonly signer: NativeExecutionReactorGrantSigner;
  readonly ttlMs?: number;
}): Promise<GnosisExecutionAuthContext | null> {
  if (!input.executionAuth) {
    return null;
  }
  if (
    !input.runtimeBindingHash ||
    input.runtimeBindingHash.trim().length === 0
  ) {
    return input.executionAuth;
  }
  if (
    !input.executionAuth.principal ||
    input.executionAuth.principal.trim().length === 0
  ) {
    return input.executionAuth;
  }

  const now = Date.now();
  const expiresAt = now + Math.max(1_000, input.ttlMs ?? 60_000);
  const resource = `${input.method.toUpperCase()} ${input.path}`;
  const ucanCapabilities =
    input.executionAuth.capabilities.map(serializeCapability);
  const requestGrant = await createReactorHandleGrant({
    claims: {
      handleId: `request:${globalThis.crypto.randomUUID()}`,
      kind: 'request',
      ownerDid: input.executionAuth.principal,
      audience: input.audience,
      resource,
      operations: ['request.metadata', 'request.read'],
      ucanCapabilities,
      runtimeBindingHash: input.runtimeBindingHash,
      issuedAt: now,
      expiresAt,
    },
    privateKey: input.signer.privateKey,
    keyId: input.signer.keyId,
    publicKey: input.signer.publicKey,
  });
  const responseGrant = await createReactorHandleGrant({
    claims: {
      handleId: `response:${globalThis.crypto.randomUUID()}`,
      kind: 'response',
      ownerDid: input.executionAuth.principal,
      audience: input.audience,
      resource: `${resource}#response`,
      operations: ['response.start', 'response.write', 'response.end'],
      ucanCapabilities,
      runtimeBindingHash: input.runtimeBindingHash,
      issuedAt: now,
      expiresAt,
    },
    privateKey: input.signer.privateKey,
    keyId: input.signer.keyId,
    publicKey: input.signer.publicKey,
  });

  return normalizeExecutionAuthContext({
    ...input.executionAuth,
    reactorHandleGrants: [
      ...(input.executionAuth.reactorHandleGrants ?? []),
      requestGrant,
      responseGrant,
    ] as ReactorHandleGrantEnvelope[],
  });
}
