// @a0n/.ignored_gnosis/src/native-bridge.ts
import { spawn } from 'node:child_process';
import { join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { 
  type ManifoldExecutor, 
  type ManifoldResult 
} from './manifold-executor';
import { type ChaosConfig } from './void-game-types';

const __dirname = fileURLToPath(new URL('.', import.meta.url));

/**
 * NativeManifoldExecutor bridges the TypeScript Gnosis ecosystem
 * with the high-performance Rust 'GnosisRayon' swarm engine.
 * 
 * It utilizes the `chaos-swarm` binary from `gnosis-chaos` as a sub-process,
 * passing configuration via JSON and harvesting swarm metrics.
 */
export class NativeManifoldExecutor implements ManifoldExecutor {
  private readonly binaryPath: string;

  constructor(customBinaryPath?: string) {
    this.binaryPath =
      customBinaryPath ??
      join(__dirname, '../gnosis-chaos/target/release/chaos-swarm');
  }

  async execute(
    seed: bigint,
    rounds: number,
    config: ChaosConfig
  ): Promise<ManifoldResult> {
    const input = {
      seed: Number(seed), // Rust expects u64, we use Number if safe or string
      rounds,
      config
    };

    return new Promise((resolve, reject) => {
      const child = spawn(this.binaryPath, ['--ffi-json', JSON.stringify(input)]);
      
      let stdout = '';
      let stderr = '';

      child.stdout.on('data', (data) => { stdout += data; });
      child.stderr.on('data', (data) => { stderr += data; });

      child.on('close', (code) => {
        if (code !== 0) {
          reject(new Error(`Native swarm engine exited with code ${code}: ${stderr}`));
          return;
        }

        try {
          const output = JSON.parse(stdout);
          resolve({
            durationMs: output.durationMs,
            metrics: output.metrics,
            activeSurvivors: output.activeSurvivors,
            survivalRate: output.activeSurvivors / config.agentCount,
            clinamenPressure: output.clinamenPressure
          });
        } catch (err) {
          reject(new Error(`Failed to parse native swarm output: ${err}\nRaw output: ${stdout}`));
        }
      });
    });
  }
}
