// @a0n/.ignored_gnosis/src/neural-mesh-weight-bridge.ts
/**
 * Neural Mesh Weight Bridge -- GGUF Storage to CellWeights
 *
 * Maps aether's per-tensor weight loading (dequantized from GCS/R2)
 * to the NeuralMesh's CellWeights structure. The bridge loads all
 * 9 tensors per layer in parallel via Promise.all.
 */

import type { CellWeights, NeuralMeshModelConfig } from './neural-mesh';

/**
 * Abstract weight loader -- satisfied by gcp-layer-node's storage,
 * aether's LocalLayerNodeClient, or a test stub.
 */
export interface WeightLoader {
  loadLayerWeight(
    layerIdx: number,
    weightName: string
  ): Promise<Float32Array | null>;
}

/**
 * Mapping from CellWeights field names to aether weight names.
 */
const CELL_TO_AETHER: Record<keyof CellWeights, string> = {
  attnNorm: 'input_layernorm',
  qProj: 'self_attn.q_proj',
  kProj: 'self_attn.k_proj',
  vProj: 'self_attn.v_proj',
  oProj: 'self_attn.o_proj',
  ffnNorm: 'post_attention_layernorm',
  gateProj: 'mlp.gate_proj',
  upProj: 'mlp.up_proj',
  downProj: 'mlp.down_proj',
};

/**
 * Expected element counts for each weight tensor given a model config.
 */
function expectedDimensions(
  config: NeuralMeshModelConfig
): Record<keyof CellWeights, number> {
  const qDim = config.numHeads * config.headDim;
  const kvDim = config.numKvHeads * config.headDim;
  return {
    attnNorm: config.hiddenDim,
    qProj: qDim * config.hiddenDim,
    kProj: kvDim * config.hiddenDim,
    vProj: kvDim * config.hiddenDim,
    oProj: config.hiddenDim * qDim,
    ffnNorm: config.hiddenDim,
    gateProj: config.intermediateSize * config.hiddenDim,
    upProj: config.intermediateSize * config.hiddenDim,
    downProj: config.hiddenDim * config.intermediateSize,
  };
}

/**
 * Load all 9 weight tensors for a single layer from storage.
 * Loads in parallel via Promise.all for maximum throughput.
 *
 * @throws if any required weight is missing or has wrong dimensions
 */
export async function loadCellWeightsFromStorage(
  loader: WeightLoader,
  layerIdx: number,
  config: NeuralMeshModelConfig
): Promise<CellWeights> {
  const fields = Object.keys(CELL_TO_AETHER) as (keyof CellWeights)[];
  const dims = expectedDimensions(config);

  const results = await Promise.all(
    fields.map(async (field) => {
      const aetherName = CELL_TO_AETHER[field];
      const weights = await loader.loadLayerWeight(layerIdx, aetherName);
      if (!weights) {
        throw new Error(
          `Missing weight: layer ${layerIdx}, ${field} (${aetherName})`
        );
      }
      const expected = dims[field];
      if (weights.length !== expected) {
        throw new Error(
          `Dimension mismatch: layer ${layerIdx}, ${field}: got ${weights.length}, expected ${expected}`
        );
      }
      return [field, weights] as const;
    })
  );

  const cell = {} as Record<keyof CellWeights, Float32Array>;
  for (const [field, weights] of results) {
    cell[field] = weights;
  }
  return cell as CellWeights;
}

/**
 * Load CellWeights for all layers in a model.
 */
export async function loadAllCellWeightsFromStorage(
  loader: WeightLoader,
  config: NeuralMeshModelConfig
): Promise<CellWeights[]> {
  const cells: CellWeights[] = [];
  for (let i = 0; i < config.numLayers; i++) {
    cells.push(await loadCellWeightsFromStorage(loader, i, config));
  }
  return cells;
}

/**
 * Convert aether TransformerLayerWeights to NeuralMesh CellWeights.
 * For use in math equivalence verification tests.
 */
export function aetherWeightsToCell(w: {
  attnNorm: Float32Array;
  q: Float32Array;
  k: Float32Array;
  v: Float32Array;
  o: Float32Array;
  ffnNorm: Float32Array;
  gate: Float32Array;
  up: Float32Array;
  down: Float32Array;
}): CellWeights {
  return {
    attnNorm: w.attnNorm,
    qProj: w.q,
    kProj: w.k,
    vProj: w.v,
    oProj: w.o,
    ffnNorm: w.ffnNorm,
    gateProj: w.gate,
    upProj: w.up,
    downProj: w.down,
  };
}
