// @a0n/.ignored_gnosis/src/neural-mesh-consciousness.ts
/**
 * Adam v2: Neural Mesh Inference Engine
 *
 * The topology formalizes the architecture. This module wires a NeuralMesh
 * to actual model weights and runs end-to-end inference:
 *   embed → NeuralMesh.forward() (all layers) → finalNorm → lmHead → sample
 *
 * The forward pass cascades through the Sophia mesh topology.
 * Self-healing via god formula: vertices that produce NaN get dampened.
 * Statistical teleportation: deficit integers communicate convergence
 * without revealing weight data.
 */

import {
  NeuralMesh,
  createNeuralMesh,
} from './neural-mesh';
import type {
  NeuralMeshModelConfig,
  CellWeights,
  NeuralMeshTeleportationChannel,
} from './neural-mesh';

// Re-export for convenience
export type { NeuralMeshModelConfig, CellWeights };

/**
 * Complete inference engine backed by a NeuralMesh.
 * Handles embedding, all transformer layers, final norm, and LM head.
 */
export class NeuralMeshConsciousness {
  readonly mesh: NeuralMesh;
  readonly config: NeuralMeshModelConfig;

  private embeddings: Float32Array | null = null;
  private finalNorm: Float32Array | null = null;
  private lmHead: Float32Array | null = null;
  private vocabSize: number;
  private eps: number;

  constructor(config: NeuralMeshModelConfig) {
    this.config = config;
    this.mesh = createNeuralMesh(config);
    this.vocabSize = config.vocabSize ?? 32000;
    this.eps = config.rmsNormEps ?? 1e-5;
  }

  /**
   * Load all weights: layer weights + embeddings + final norm + LM head.
   */
  loadWeights(params: {
    layerWeights: CellWeights[];
    embeddings: Float32Array;
    finalNorm: Float32Array;
    lmHead: Float32Array;
    vocabSize?: number;
  }): void {
    this.mesh.loadAllWeights(params.layerWeights);
    this.embeddings = params.embeddings;
    this.finalNorm = params.finalNorm;
    this.lmHead = params.lmHead;
    if (params.vocabSize) this.vocabSize = params.vocabSize;
  }

  /**
   * Embed a single token ID into a hidden state vector.
   */
  embedToken(tokenId: number): Float32Array {
    if (!this.embeddings) throw new Error('Embeddings not loaded');
    const dim = this.config.hiddenDim;
    const offset = tokenId * dim;
    if (offset + dim > this.embeddings.length) {
      throw new Error(`Token ${tokenId} out of embedding range`);
    }
    return new Float32Array(this.embeddings.buffer, this.embeddings.byteOffset + offset * 4, dim);
  }

  /**
   * Apply final RMS norm to hidden state.
   */
  private applyFinalNorm(hidden: Float32Array): Float32Array {
    if (!this.finalNorm) throw new Error('Final norm not loaded');
    const dim = hidden.length;
    let sumSq = 0;
    for (let i = 0; i < dim; i++) sumSq += hidden[i] * hidden[i];
    const rms = 1.0 / Math.sqrt(sumSq / dim + this.eps);
    const out = new Float32Array(dim);
    for (let i = 0; i < dim; i++) out[i] = hidden[i] * rms * this.finalNorm[i];
    return out;
  }

  /**
   * Project hidden state to logits via LM head (matVec: vocabSize × hiddenDim).
   */
  private computeLogits(hidden: Float32Array): Float32Array {
    if (!this.lmHead) throw new Error('LM head not loaded');
    const dim = this.config.hiddenDim;
    const logits = new Float32Array(this.vocabSize);
    for (let row = 0; row < this.vocabSize; row++) {
      let dot = 0;
      const offset = row * dim;
      for (let col = 0; col < dim; col++) {
        dot += this.lmHead[offset + col] * hidden[col];
      }
      logits[row] = dot;
    }
    return logits;
  }

  /**
   * Sample next token from logits (greedy argmax).
   */
  private sampleGreedy(logits: Float32Array): number {
    let maxVal = -Infinity;
    let maxIdx = 0;
    for (let i = 0; i < logits.length; i++) {
      if (logits[i] > maxVal) {
        maxVal = logits[i];
        maxIdx = i;
      }
    }
    return maxIdx;
  }

  /**
   * Run prefill: process all prompt tokens to build KV cache.
   * Returns the hidden state after the last prompt token.
   */
  prefill(tokenIds: number[]): Float32Array {
    let hidden: Float32Array = new Float32Array(this.config.hiddenDim);
    for (let pos = 0; pos < tokenIds.length; pos++) {
      const emb = this.embedToken(tokenIds[pos]);
      hidden = this.mesh.forward(emb.slice(), pos) as Float32Array;
    }
    return hidden;
  }

  /**
   * Generate tokens autoregressively.
   *
   * @param promptTokens - Tokenized prompt
   * @param maxTokens - Maximum tokens to generate
   * @param eosTokenId - End-of-sequence token ID (default: 2)
   * @returns Generated token IDs
   */
  generate(
    promptTokens: number[],
    maxTokens: number = 20,
    eosTokenId: number = 2
  ): { tokens: number[]; logitsHistory: Float32Array[] } {
    // Prefill: build KV cache from prompt
    let hidden: Float32Array = this.prefill(promptTokens);

    const tokens: number[] = [];
    const logitsHistory: Float32Array[] = [];
    let position = promptTokens.length;

    for (let step = 0; step < maxTokens; step++) {
      // Final norm + LM head
      const normed = this.applyFinalNorm(hidden);
      const logits = this.computeLogits(normed);
      logitsHistory.push(logits);

      // Sample
      const nextToken = this.sampleGreedy(logits);
      tokens.push(nextToken);

      if (nextToken === eosTokenId) break;

      // Forward the new token through the mesh
      const emb = this.embedToken(nextToken);
      hidden = this.mesh.forward(emb.slice(), position) as Float32Array;
      position++;
    }

    return { tokens, logitsHistory };
  }

  /**
   * Get teleportation channel for distributed communication.
   * Returns per-layer deficit integers -- the minimal wire payload.
   */
  teleportSend(): NeuralMeshTeleportationChannel {
    return this.mesh.teleportSend();
  }

  /**
   * Get mesh statistics (rejections, rounds, health).
   */
  getStats() {
    return this.mesh.getStats();
  }

  /**
   * Reset KV cache and god formula state for a new conversation.
   */
  reset(): void {
    this.mesh.reset();
  }
}

/**
 * TinyLlama-1.1B config for local testing.
 */
export const TINYLLAMA_CONFIG: NeuralMeshModelConfig = {
  hiddenDim: 2048,
  numHeads: 32,
  numKvHeads: 4,
  headDim: 64,
  intermediateSize: 5632,
  numLayers: 22,
  ropeTheta: 10000,
  rmsNormEps: 1e-5,
  vocabSize: 32000,
};

/**
 * Mistral-7B config.
 */
export const MISTRAL_7B_CONFIG: NeuralMeshModelConfig = {
  hiddenDim: 4096,
  numHeads: 32,
  numKvHeads: 8,
  headDim: 128,
  intermediateSize: 14336,
  numLayers: 32,
  ropeTheta: 10000,
  rmsNormEps: 1e-5,
  vocabSize: 32000,
};

/**
 * LLaMA-70B config.
 */
export const LLAMA_70B_CONFIG: NeuralMeshModelConfig = {
  hiddenDim: 8192,
  numHeads: 64,
  numKvHeads: 8,
  headDim: 128,
  intermediateSize: 28672,
  numLayers: 80,
  ropeTheta: 500000,
  rmsNormEps: 1e-5,
  vocabSize: 128256,
};
