// @a0n/.ignored_gnosis/src/neural-mesh.ts
/**
 * Neural Mesh -- Weights as Vertices, Topology as Architecture
 *
 * Adam v2. The mesh vertices DO the computation. Each vertex holds a
 * real weight tensor and runs a real matVec. The forward pass formalizes the
 * cascade through the spiderweb topology.
 *
 * Architecture:
 *   - One cell per transformer layer
 *   - Nine vertices per cell (k=3, s=3 → Sophia β₁=9)
 *   - Each vertex holds one weight tensor: attnNorm, Q, K, V, O, ffnNorm, gate, up, down
 *   - Ring 0 (3 radials): attnNorm, Q-proj, ffnNorm — consume raw hidden state
 *   - Ring 1 (3 radials): K-proj, V-proj, gate-proj — consume normed outputs
 *   - Ring 2 (3 radials): O-proj, up-proj, down-proj
 *   - Belt edges: output collection / residual addition
 *   - Hub: input hidden state
 *   - Endpoints: output hidden state → next cell's hub
 *
 * Self-healing via god formula: vertices that produce NaN get rejected →
 * lower weight → lower contribution. The mesh dampens unstable layers.
 *
 * @packageDocumentation
 */

import { godFormula } from './holy-trinity';
import { SchedulingMesh, type MeshVertex, VOID, FORK, VENT } from './scheduling-mesh';
import {
  type GnosticMeshLevel,
  GNOSTIC_MESHES,
  GNOSTIC_BETA1,
  resolveSyzygyDimension,
  compositeDimension,
} from './spiderweb-topology';

// ═══════════════════════════════════════════════════════════════════════
// TYPES
// ═══════════════════════════════════════════════════════════════════════

/**
 * Model configuration for a standard transformer.
 * Compatible with LLaMA / Mistral / TinyLlama architectures.
 */
export interface NeuralMeshModelConfig {
  /** Hidden dimension (e.g. 2048 for TinyLlama, 4096 for LLaMA-7B) */
  readonly hiddenDim: number;
  /** Number of attention heads (e.g. 32) */
  readonly numHeads: number;
  /** Number of KV heads for GQA (e.g. 4 for TinyLlama, 32 for LLaMA-7B) */
  readonly numKvHeads: number;
  /** Head dimension (usually hiddenDim / numHeads) */
  readonly headDim: number;
  /** Intermediate size for SwiGLU MLP (e.g. 5632 for TinyLlama) */
  readonly intermediateSize: number;
  /** Number of transformer layers */
  readonly numLayers: number;
  /** RoPE theta (default 10000) */
  readonly ropeTheta?: number;
  /** RMS norm epsilon (default 1e-5) */
  readonly rmsNormEps?: number;
  /** Vocabulary size (for output logits) */
  readonly vocabSize?: number;
}

/**
 * The nine weight tensors held by a single transformer layer cell.
 * Each maps to one vertex in the Sophia mesh.
 */
export interface CellWeights {
  /** Ring 0, Radial 0: Attention RMS norm [hiddenDim] */
  attnNorm: Float32Array;
  /** Ring 0, Radial 1: Query projection [qDim × hiddenDim] */
  qProj: Float32Array;
  /** Ring 0, Radial 2: FFN RMS norm [hiddenDim] */
  ffnNorm: Float32Array;
  /** Ring 1, Radial 0: Key projection [kvDim × hiddenDim] */
  kProj: Float32Array;
  /** Ring 1, Radial 1: Value projection [kvDim × hiddenDim] */
  vProj: Float32Array;
  /** Ring 1, Radial 2: Gate projection [intSize × hiddenDim] */
  gateProj: Float32Array;
  /** Ring 2, Radial 0: Output projection [hiddenDim × qDim] */
  oProj: Float32Array;
  /** Ring 2, Radial 1: Up projection [intSize × hiddenDim] */
  upProj: Float32Array;
  /** Ring 2, Radial 2: Down projection [hiddenDim × intSize] */
  downProj: Float32Array;
}

/**
 * Per-layer key-value cache for autoregressive generation.
 */
export interface NeuralMeshKVCache {
  /** Cached key vectors per position [seqLen][kvDim] */
  keys: Float32Array[];
  /** Cached value vectors per position [seqLen][kvDim] */
  values: Float32Array[];
}

/**
 * Stats from a neural mesh forward pass.
 */
export interface NeuralMeshForwardStats {
  /** Number of layers processed */
  layersProcessed: number;
  /** Number of NaN rejections (god formula dampened) */
  nanRejections: number;
  /** Total god formula rejections across all vertices */
  totalRejections: number;
  /** Per-layer vertex weights (9 per layer) */
  vertexWeights: number[][];
  /** Time taken in ms */
  elapsedMs: number;
}

/**
 * The vertex-to-weight mapping within a cell.
 * Sophia mesh: k=3 radials, s=3 rings → 9 relay vertices + 1 hub + 3 endpoints = 13 per cell.
 * We use the 9 relay vertices for the 9 weight tensors.
 */
const WEIGHT_NAMES = [
  'attnNorm', 'qProj', 'ffnNorm',  // Ring 0 (3 radials)
  'kProj',    'vProj', 'gateProj',  // Ring 1 (3 radials)
  'oProj',    'upProj', 'downProj', // Ring 2 (3 radials)
] as const;

type WeightName = typeof WEIGHT_NAMES[number];

// ═══════════════════════════════════════════════════════════════════════
// MATH OPERATIONS (inlined for zero-dependency neural mesh)
// ═══════════════════════════════════════════════════════════════════════
// These mirror the aether engine ops but are self-contained so the
// neural mesh can operate without importing cross-package dependencies.

/** Matrix × vector: result[i] = Σ_j matrix[i*cols+j] * vector[j] */
function matVec(
  matrix: Float32Array,
  vector: Float32Array,
  rows: number,
  cols: number,
): Float32Array {
  const result = new Float32Array(rows);
  const cols4 = cols & ~3;
  for (let i = 0; i < rows; i++) {
    const offset = i * cols;
    let s0 = 0, s1 = 0, s2 = 0, s3 = 0;
    for (let j = 0; j < cols4; j += 4) {
      s0 += readArrayValue(matrix, offset + j, 'matVec matrix value') * readArrayValue(vector, j, 'matVec vector value');
      s1 += readArrayValue(matrix, offset + j + 1, 'matVec matrix value') * readArrayValue(vector, j + 1, 'matVec vector value');
      s2 += readArrayValue(matrix, offset + j + 2, 'matVec matrix value') * readArrayValue(vector, j + 2, 'matVec vector value');
      s3 += readArrayValue(matrix, offset + j + 3, 'matVec matrix value') * readArrayValue(vector, j + 3, 'matVec vector value');
    }
    let sum = s0 + s1 + s2 + s3;
    for (let j = cols4; j < cols; j++) {
      sum += readArrayValue(matrix, offset + j, 'matVec matrix tail') * readArrayValue(vector, j, 'matVec vector tail');
    }
    result[i] = sum;
  }
  return result;
}

/** RMS normalization: x * weight / rms(x). */
function rmsNorm(
  x: Float32Array,
  weight: Float32Array,
  eps: number,
): Float32Array {
  const len = x.length;
  let sumSq = 0;
  for (const value of x) sumSq += value * value;
  const invRms = 1 / Math.sqrt(sumSq / len + eps);
  const result = new Float32Array(len);
  for (let i = 0; i < len; i++) {
    const xValue = readArrayValue(x, i, 'rmsNorm input value');
    const weightValue = readArrayValue(weight, i, 'rmsNorm weight value');
    result[i] = xValue * invRms * weightValue;
  }
  return result;
}

/** Softmax with numerical stability. */
function softmax(x: Float32Array): Float32Array {
  const len = x.length;
  let max = Number.NEGATIVE_INFINITY;
  for (const value of x) if (value > max) max = value;
  const result = new Float32Array(len);
  let sum = 0;
  for (let i = 0; i < len; i++) {
    const value = readArrayValue(x, i, 'softmax input value');
    const e = Math.exp(Math.min(value - max, 80));
    result[i] = e;
    sum += e;
  }
  if (sum > 0 && isFinite(sum)) {
    const inv = 1 / sum;
    for (let i = 0; i < len; i++) {
      const current = result[i];
      if (current === undefined) {
        throw new Error('softmax: missing result value');
      }
      result[i] = current * inv;
    }
  } else {
    const u = 1 / len;
    for (let i = 0; i < len; i++) result[i] = u;
  }
  return result;
}

/** Fused SiLU(gate) * up in a single pass (SwiGLU). */
function fusedSiluMul(gate: Float32Array, up: Float32Array): Float32Array {
  const len = gate.length;
  const result = new Float32Array(len);
  for (let i = 0; i < len; i++) {
    const g = readArrayValue(gate, i, 'fusedSiluMul gate value');
    const u = readArrayValue(up, i, 'fusedSiluMul up value');
    result[i] = (g / (1 + Math.exp(-g))) * u;
  }
  return result;
}

/** Element-wise add: a + b. */
function add(a: Float32Array, b: Float32Array): Float32Array {
  const result = new Float32Array(a.length);
  for (let i = 0; i < a.length; i++) {
    const aValue = readArrayValue(a, i, 'add left value');
    const bValue = readArrayValue(b, i, 'add right value');
    result[i] = aValue + bValue;
  }
  return result;
}

function readArrayValue(values: ArrayLike<number>, index: number, context: string): number {
  const value = values[index];
  if (value === undefined) {
    throw new Error(context);
  }
  return value;
}

/** RoPE frequency table cache. */
const _ropeFreqCache = new Map<string, Float64Array>();

function getRopeFreqs(headDim: number, theta: number): Float64Array {
  const key = `${headDim}:${theta}`;
  let freqs = _ropeFreqCache.get(key);
  if (freqs) return freqs;
  const halfDim = headDim / 2;
  freqs = new Float64Array(halfDim);
  for (let k = 0; k < halfDim; k++) {
    freqs[k] = 1.0 / Math.pow(theta, (2 * k) / headDim);
  }
  _ropeFreqCache.set(key, freqs);
  return freqs;
}

/** Apply Rotary Position Embedding. */
function applyRoPE(
  x: Float32Array,
  position: number,
  headDim: number,
  numHeads: number,
  theta: number,
): Float32Array {
  const result = new Float32Array(x.length);
  const halfDim = headDim / 2;
  const freqs = getRopeFreqs(headDim, theta);
  const cosVals = new Float64Array(halfDim);
  const sinVals = new Float64Array(halfDim);
  for (let k = 0; k < halfDim; k++) {
    const angle = position * readArrayValue(freqs, k, 'applyRoPE frequency');
    cosVals[k] = Math.cos(angle);
    sinVals[k] = Math.sin(angle);
  }
  for (let h = 0; h < numHeads; h++) {
    const start = h * headDim;
    for (let k = 0; k < halfDim; k++) {
      const i = k * 2;
      const x0 = readArrayValue(x, start + i, 'applyRoPE x0');
      const x1 = readArrayValue(x, start + i + 1, 'applyRoPE x1');
      const cosValue = readArrayValue(cosVals, k, 'applyRoPE cos value');
      const sinValue = readArrayValue(sinVals, k, 'applyRoPE sin value');
      result[start + i] = x0 * cosValue - x1 * sinValue;
      result[start + i + 1] = x0 * sinValue + x1 * cosValue;
    }
  }
  return result;
}

/** Check if a Float32Array contains any NaN. */
function hasNaN(arr: Float32Array): boolean {
  for (const value of arr) {
    if (!isFinite(value)) return true;
  }
  return false;
}

// ═══════════════════════════════════════════════════════════════════════
// NEURAL MESH
// ═══════════════════════════════════════════════════════════════════════

/**
 * NeuralMesh — Weights as Vertices, Topology as Architecture.
 *
 * A Sophia mesh (k=3, s=3) where each cell is one transformer layer,
 * each vertex holds one weight tensor, and the forward pass cascades
 * through the mesh topology.
 *
 * This is Adam v2. Not a scheduler. The mesh vertices DO the computation.
 */
export class NeuralMesh {
  /** The underlying scheduling mesh (Sophia topology). */
  readonly mesh: SchedulingMesh;
  /** Model configuration. */
  readonly config: NeuralMeshModelConfig;
  /** Per-layer weights. */
  private layerWeights: CellWeights[];
  /** Per-layer KV cache. */
  private kvCaches: NeuralMeshKVCache[];
  /** Derived dimensions. */
  private readonly qDim: number;
  private readonly kvDim: number;
  private readonly intSize: number;
  private readonly headDim: number;
  private readonly eps: number;
  private readonly theta: number;
  /** God formula round counter. */
  private rounds: number;
  /** Per-vertex rejection counts (flattened: [layer * 9 + vertexIdx]). */
  private rejections: number[];
  /** Per-vertex weights (flattened: [layer * 9 + vertexIdx]). */
  private vertexGodWeights: number[];

  constructor(config: NeuralMeshModelConfig) {
    this.config = config;
    this.headDim = config.headDim;
    this.qDim = config.numHeads * config.headDim;
    this.kvDim = config.numKvHeads * config.headDim;
    this.intSize = config.intermediateSize;
    this.eps = config.rmsNormEps ?? 1e-5;
    this.theta = config.ropeTheta ?? 10000.0;
    this.rounds = 0;

    // Build a Sophia mesh with one cell per transformer layer
    this.mesh = new SchedulingMesh('sophia', config.numLayers);

    // Initialize empty weights and KV caches
    this.layerWeights = [];
    this.kvCaches = [];
    this.rejections = new Array(config.numLayers * 9).fill(0);
    this.vertexGodWeights = new Array(config.numLayers * 9).fill(1);

    for (let i = 0; i < config.numLayers; i++) {
      this.kvCaches.push({ keys: [], values: [] });
    }
  }

  /** Number of cells (layers) in the mesh. */
  get numLayers(): number {
    return this.config.numLayers;
  }

  /** Beta1 per cell (always 9 for Sophia). */
  get beta1(): number {
    return 9;
  }

  /** Composite dimension across all layers. */
  get compositeDimension(): number {
    return compositeDimension(this.numLayers, this.beta1);
  }

  /** Whether all layers have weights loaded. */
  get isLoaded(): boolean {
    return this.layerWeights.length === this.numLayers;
  }

  // ─── Weight Loading ─────────────────────────────────────────────

  /**
   * Load weight tensors for a specific layer.
   * Each weight is loaded into its corresponding mesh vertex.
   */
  loadLayerWeights(layerIdx: number, weights: CellWeights): void {
    if (layerIdx < 0 || layerIdx >= this.numLayers) {
      throw new Error(`Layer index ${layerIdx} out of range [0, ${this.numLayers})`);
    }

    // Validate dimensions
    this.validateWeightDimensions(weights);

    // Extend array if needed (layers can load out of order)
    while (this.layerWeights.length <= layerIdx) {
      this.layerWeights.push(null as unknown as CellWeights);
    }
    this.layerWeights[layerIdx] = weights;

    // Store weights as payloads on the mesh vertices
    const vertices = this.mesh.cellVertices(layerIdx);
    const relays = vertices.filter(v => v.type === 'relay');

    // Map the 9 relay vertices to their weight tensors
    // The relays are ordered by radial then ring: (r0,s0), (r0,s1), (r0,s2), (r1,s0), ...
    // We need: ring 0 = [attnNorm, qProj, ffnNorm], ring 1 = [kProj, vProj, gateProj], ring 2 = [oProj, upProj, downProj]
    const weightArrays: Float32Array[] = [
      weights.attnNorm, weights.qProj, weights.ffnNorm,
      weights.kProj, weights.vProj, weights.gateProj,
      weights.oProj, weights.upProj, weights.downProj,
    ];

    for (let i = 0; i < Math.min(relays.length, 9); i++) {
      const relay = relays[i];
      const weight = weightArrays[i];
      if (!relay || !weight) {
        throw new Error('loadLayerWeights: missing relay or weight');
      }
      relay.payload = weight;
    }
  }

  /**
   * Load all layer weights at once.
   */
  loadAllWeights(allWeights: CellWeights[]): void {
    if (allWeights.length !== this.numLayers) {
      throw new Error(
        `Expected ${this.numLayers} layer weight sets, got ${allWeights.length}`
      );
    }
    for (let i = 0; i < this.numLayers; i++) {
      const weights = allWeights[i];
      if (!weights) {
        throw new Error(`loadAllWeights: missing layer weights at index ${i}`);
      }
      this.loadLayerWeights(i, weights);
    }
  }

  private validateWeightDimensions(w: CellWeights): void {
    const { hiddenDim } = this.config;
    const errs: string[] = [];

    if (w.attnNorm.length !== hiddenDim)
      errs.push(`attnNorm: expected ${hiddenDim}, got ${w.attnNorm.length}`);
    if (w.ffnNorm.length !== hiddenDim)
      errs.push(`ffnNorm: expected ${hiddenDim}, got ${w.ffnNorm.length}`);
    if (w.qProj.length !== this.qDim * hiddenDim)
      errs.push(`qProj: expected ${this.qDim * hiddenDim}, got ${w.qProj.length}`);
    if (w.kProj.length !== this.kvDim * hiddenDim)
      errs.push(`kProj: expected ${this.kvDim * hiddenDim}, got ${w.kProj.length}`);
    if (w.vProj.length !== this.kvDim * hiddenDim)
      errs.push(`vProj: expected ${this.kvDim * hiddenDim}, got ${w.vProj.length}`);
    if (w.oProj.length !== this.config.hiddenDim * this.qDim)
      errs.push(`oProj: expected ${hiddenDim * this.qDim}, got ${w.oProj.length}`);
    if (w.gateProj.length !== this.intSize * hiddenDim)
      errs.push(`gateProj: expected ${this.intSize * hiddenDim}, got ${w.gateProj.length}`);
    if (w.upProj.length !== this.intSize * hiddenDim)
      errs.push(`upProj: expected ${this.intSize * hiddenDim}, got ${w.upProj.length}`);
    if (w.downProj.length !== hiddenDim * this.intSize)
      errs.push(`downProj: expected ${hiddenDim * this.intSize}, got ${w.downProj.length}`);

    if (errs.length > 0) {
      throw new Error(`Weight dimension mismatch:\n  ${errs.join('\n  ')}`);
    }
  }

  // ─── Forward Pass ──────────────────────────────────────────────

  /**
   * Forward pass: cascade hidden state through the mesh.
   *
   * Each cell (layer) processes the hidden state through its 9 vertices:
   *   Ring 0: norm + projections (parallel via fork on 3 radials)
   *   Ring 1: K/V/gate projections (consume normed outputs)
   *   Ring 2: O/up/down projections + residuals (fold at belt)
   *
   * The god formula governs computation: vertices that produce NaN
   * get rejected → lower weight → dampened contribution.
   */
  forward(hidden: Float32Array, position: number): Float32Array {
    if (!this.isLoaded) {
      throw new Error('Not all layer weights are loaded');
    }
    if (hidden.length !== this.config.hiddenDim) {
      throw new Error(
        `Input size mismatch: expected ${this.config.hiddenDim}, got ${hidden.length}`
      );
    }

    const startTime = typeof performance !== 'undefined' ? performance.now() : Date.now();
    let state = hidden;
    let nanRejections = 0;

    for (let layer = 0; layer < this.numLayers; layer++) {
      const w = this.layerWeights[layer];
      const cache = this.kvCaches[layer];
      if (!w || !cache) {
        throw new Error(`forward: missing layer state for layer ${layer}`);
      }
      const vIdx = layer * 9;

      // ─── Ring 0: attnNorm + Q + ffnNorm (parallel on same input) ───
      const normed = this.computeVertex(vIdx + 0, () =>
        rmsNorm(state, w.attnNorm, this.eps)
      );

      const q = this.computeVertex(vIdx + 1, () =>
        matVec(w.qProj, normed, this.qDim, this.config.hiddenDim)
      );

      // ffnNorm is computed early but used later (stored for Ring 2)
      // The topology allows this: Ring 0 radial 2 feeds forward to Ring 2
      const ffnNormedInput = state; // Will be re-normed after attention residual

      // ─── Ring 1: K + V + gate (consume normed output) ───
      const k = this.computeVertex(vIdx + 3, () =>
        matVec(w.kProj, normed, this.kvDim, this.config.hiddenDim)
      );

      const v = this.computeVertex(vIdx + 4, () =>
        matVec(w.vProj, normed, this.kvDim, this.config.hiddenDim)
      );

      // Apply RoPE
      const ropeQ = applyRoPE(q, position, this.headDim, this.config.numHeads, this.theta);
      const ropeK = applyRoPE(k, position, this.headDim, this.config.numKvHeads, this.theta);

      // Cache K, V
      cache.keys.push(new Float32Array(ropeK));
      cache.values.push(new Float32Array(v));

      // ─── Attention computation at spiral junction ───
      const seqLen = cache.keys.length;
      const scale = 1.0 / Math.sqrt(this.headDim);
      const headsPerKvGroup = this.config.numHeads / this.config.numKvHeads;
      const attnOut = new Float32Array(this.qDim);

      for (let h = 0; h < this.config.numHeads; h++) {
        const qStart = h * this.headDim;
        const kvHead = Math.floor(h / headsPerKvGroup);
        const kvStart = kvHead * this.headDim;
        const qHead = ropeQ.subarray(qStart, qStart + this.headDim);

        // QK^T dot products
        const scores = new Float32Array(seqLen);
        for (let s = 0; s < seqLen; s++) {
          let dot = 0;
          const keyRow = cache.keys[s];
          if (!keyRow) {
            throw new Error('forward: missing cached key row');
          }
          for (let d = 0; d < this.headDim; d++) {
            const qValue = readArrayValue(qHead, d, 'forward qHead value');
            const keyValue = readArrayValue(keyRow, kvStart + d, 'forward cached key value');
            dot += qValue * keyValue;
          }
          scores[s] = dot * scale;
        }

        // Softmax
        const attnWeights = softmax(scores);

        // Weighted sum of values
        for (let d = 0; d < this.headDim; d++) {
          let sum = 0;
          for (let s = 0; s < seqLen; s++) {
            const attnWeight = readArrayValue(attnWeights, s, 'forward attention weight');
            const valueRow = cache.values[s];
            if (!valueRow) {
              throw new Error('forward: missing cached value row');
            }
            const cachedValue = readArrayValue(valueRow, kvStart + d, 'forward cached value');
            sum += attnWeight * cachedValue;
          }
          attnOut[qStart + d] = sum;
        }
      }

      // ─── Ring 2: O-proj + residual, up-proj, down-proj + residual ───
      const projected = this.computeVertex(vIdx + 6, () =>
        matVec(w.oProj, attnOut, this.config.hiddenDim, this.qDim)
      );

      const residual1 = add(state, projected);

      // FFN: norm + SwiGLU
      const ffnNormed = this.computeVertex(vIdx + 2, () =>
        rmsNorm(residual1, w.ffnNorm, this.eps)
      );

      const gateOut = this.computeVertex(vIdx + 5, () =>
        matVec(w.gateProj, ffnNormed, this.intSize, this.config.hiddenDim)
      );

      const upOut = this.computeVertex(vIdx + 7, () =>
        matVec(w.upProj, ffnNormed, this.intSize, this.config.hiddenDim)
      );

      const swiglu = fusedSiluMul(gateOut, upOut);

      const downOut = this.computeVertex(vIdx + 8, () =>
        matVec(w.downProj, swiglu, this.config.hiddenDim, this.intSize)
      );

      // Final residual
      state = add(residual1, downOut);

      // ─── God formula self-healing ───
      // If the output state contains NaN, this layer produced divergence.
      // Reject all vertices in this cell, dampening their weights.
      if (hasNaN(state)) {
        nanRejections++;
        this.rounds++;
        for (let vi = 0; vi < 9; vi++) {
          const rejectionIndex = vIdx + vi;
          const rejectionCount = this.rejections[rejectionIndex];
          if (rejectionCount === undefined) {
            throw new Error('forward: missing rejection counter');
          }
          this.rejections[rejectionIndex] = rejectionCount + 1;
          const updatedRejectionCount = this.rejections[rejectionIndex];
          if (updatedRejectionCount === undefined) {
            throw new Error('forward: missing updated rejection counter');
          }
          this.vertexGodWeights[rejectionIndex] = godFormula(
            this.rounds,
            updatedRejectionCount
          );
        }
        // Recover: revert to the input state for this layer
        state = hidden;
      }
    }

    return state;
  }

  /**
   * Execute a vertex computation with god formula modulation.
   *
   * If the vertex has a low god formula weight (from prior rejections),
   * its output magnitude is scaled down proportionally. This is the
   * self-healing mechanism: problematic weight matrices get dampened.
   */
  private computeVertex(
    flatIdx: number,
    compute: () => Float32Array,
  ): Float32Array {
    const result = compute();
    const weight = this.vertexGodWeights[flatIdx];
    if (weight === undefined) {
      throw new Error(`computeVertex: missing god weight for vertex ${flatIdx}`);
    }

    // God formula modulation: scale output by vertex weight
    // weight = R - min(v, R) + 1. At full health, weight = R + 1 (max).
    // After rejection, weight decreases toward 1 (the sliver).
    if (weight < this.rounds + 1 && this.rounds > 0) {
      const scale = weight / (this.rounds + 1);
      for (let i = 0; i < result.length; i++) {
        const current = result[i];
        if (current === undefined) {
          throw new Error('computeVertex: missing result value');
        }
        result[i] = current * scale;
      }
    }

    return result;
  }

  // ─── Cache Management ──────────────────────────────────────────

  /** Reset all KV caches (for new generation). */
  resetCache(): void {
    for (const cache of this.kvCaches) {
      cache.keys.length = 0;
      cache.values.length = 0;
    }
  }

  /** Reset god formula state (all vertices back to full weight). */
  resetGodFormula(): void {
    this.rounds = 0;
    this.rejections.fill(0);
    this.vertexGodWeights.fill(1);
  }

  /** Full reset: cache + god formula. */
  reset(): void {
    this.resetCache();
    this.resetGodFormula();
  }

  // ─── Diagnostics ───────────────────────────────────────────────

  /**
   * Get the current vertex weights across the mesh.
   * Returns a 2D array: [layer][vertexIdx] → god formula weight.
   */
  getVertexWeights(): number[][] {
    const result: number[][] = [];
    for (let layer = 0; layer < this.numLayers; layer++) {
      const layerWeights: number[] = [];
      for (let vi = 0; vi < 9; vi++) {
        const weight = this.vertexGodWeights[layer * 9 + vi];
        if (weight === undefined) {
          throw new Error('getVertexWeights: missing vertex weight');
        }
        layerWeights.push(weight);
      }
      result.push(layerWeights);
    }
    return result;
  }

  /**
   * Get forward pass statistics.
   */
  getStats(): {
    numLayers: number;
    beta1PerLayer: number;
    compositeDimension: number;
    totalVertices: number;
    totalRejections: number;
    rounds: number;
    isLoaded: boolean;
    cachePositions: number[];
  } {
    return {
      numLayers: this.numLayers,
      beta1PerLayer: this.beta1,
      compositeDimension: this.compositeDimension,
      totalVertices: this.numLayers * 9,
      totalRejections: this.rejections.reduce((s, r) => s + r, 0),
      rounds: this.rounds,
      isLoaded: this.isLoaded,
      cachePositions: this.kvCaches.map(c => c.keys.length),
    };
  }

  /**
   * Inject NaN into a specific vertex's weights for testing self-healing.
   * Returns the vertex's god formula weight before injection.
   */
  injectNaN(layerIdx: number, weightName: WeightName): number {
    if (layerIdx < 0 || layerIdx >= this.numLayers) {
      throw new Error(`Layer ${layerIdx} out of range`);
    }
    const w = this.layerWeights[layerIdx];
    if (!w) {
      throw new Error(`Missing weights for layer ${layerIdx}`);
    }
    const arr = w[weightName];
    if (!arr) throw new Error(`Unknown weight: ${weightName}`);

    const nameIdx = WEIGHT_NAMES.indexOf(weightName);
    if (nameIdx < 0) {
      throw new Error(`Unknown weight: ${weightName}`);
    }
    const prevWeight = this.vertexGodWeights[layerIdx * 9 + nameIdx];
    if (prevWeight === undefined) {
      throw new Error('injectNaN: missing previous vertex weight');
    }

    // Inject NaN into the middle of the weight tensor
    const mid = Math.floor(arr.length / 2);
    arr[mid] = NaN;

    return prevWeight;
  }

  /**
   * Get the weight name for a vertex index within a cell.
   */
  static vertexWeightName(vertexIdx: number): WeightName {
    const weightName = WEIGHT_NAMES[vertexIdx];
    if (!weightName) {
      throw new Error(`vertexWeightName: invalid vertex index ${vertexIdx}`);
    }
    return weightName;
  }

  // ─── External Rejection Recording ─────────────────────────────
  //
  // For integration with existing inference pipelines that detect NaN
  // independently (e.g., gcp-layer-node.ts clamping). The external
  // caller records the rejection; the mesh tracks it via god formula
  // without running its own forward pass.

  /**
   * Record a rejection for all 9 vertices in a layer.
   * Called by external inference code when NaN/explosion is detected.
   * Updates god formula weights and increments round counter.
   */
  recordLayerRejection(layerIdx: number): void {
    this.rounds++;
    const vIdx = layerIdx * 9;
    for (let vi = 0; vi < 9; vi++) {
      const rejectionIndex = vIdx + vi;
      const rejectionCount = this.rejections[rejectionIndex];
      if (rejectionCount === undefined) {
        throw new Error('recordLayerRejection: missing rejection counter');
      }
      this.rejections[rejectionIndex] = rejectionCount + 1;
      const updatedRejectionCount = this.rejections[rejectionIndex];
      if (updatedRejectionCount === undefined) {
        throw new Error('recordLayerRejection: missing updated rejection counter');
      }
      this.vertexGodWeights[rejectionIndex] = godFormula(
        this.rounds,
        updatedRejectionCount
      );
    }
  }

  // ─── Statistical Teleportation ─────────────────────────────────
  //
  // From StatisticalTeleportation.lean:
  //   "The Bule deficit is a single natural number that encodes the
  //    entire future entropy trajectory of a Buleyean space.
  //    Certainty transfers. Data stays local."
  //
  // Each layer's 9 vertices accumulate rejections — these form a
  // void boundary. The deficit from that boundary is ONE integer
  // that tells a receiver the layer's full convergence trajectory
  // without revealing the rejection history itself.
  //
  // For distributed inference: transmit per-layer deficits, not
  // hidden states. The receiver knows when each layer will converge
  // and how dampened it is, without seeing any weight data.

  /**
   * Compute the teleportation deficit for a single layer.
   *
   * deficit = totalRejections - minRejection × 9
   *
   * This one integer encodes the entire future convergence trajectory:
   *   - deficitAtRound(k) = max(0, deficit - k)
   *   - convergenceRound = deficit
   *   - trajectory is monotonically decreasing
   *
   * Mechanized: StatisticalTeleportation.lean (zero sorry).
   */
  layerTeleportDeficit(layerIdx: number): number {
    const base = layerIdx * 9;
    let total = 0;
    const firstRejection = this.rejections[base];
    if (firstRejection === undefined) {
      throw new Error('layerTeleportDeficit: missing base rejection');
    }
    let min = firstRejection;
    for (let i = 0; i < 9; i++) {
      const r = this.rejections[base + i];
      if (r === undefined) {
        throw new Error('layerTeleportDeficit: missing rejection entry');
      }
      total += r;
      if (r < min) min = r;
    }
    return total - min * 9;
  }

  /**
   * Extract the full teleportation channel for the mesh.
   *
   * Returns one deficit integer per layer — the minimal payload
   * needed to communicate the mesh's convergence state to a
   * remote node. The weight data stays local.
   *
   * teleportation_privacy (Lean): two meshes with different
   * rejection histories but same deficit are indistinguishable
   * to the receiver.
   */
  teleportSend(): NeuralMeshTeleportationChannel {
    const layerDeficits: number[] = [];
    for (let i = 0; i < this.numLayers; i++) {
      layerDeficits.push(this.layerTeleportDeficit(i));
    }
    return {
      layerDeficits,
      numLayers: this.numLayers,
      beta1: this.beta1,
      compositeDimension: this.compositeDimension,
      rounds: this.rounds,
      transmittedAt: Date.now(),
    };
  }

  /**
   * Receive a teleportation channel and reconstruct convergence
   * trajectories. The receiver knows WHEN each layer will converge
   * and HOW dampened it is — without seeing any weight data.
   *
   * teleportation_trajectory_from_deficit (Lean):
   *   receiverKnowsTrajectory(deficit, k) = deficit - min(k, deficit)
   *
   * teleportation_convergence_round (Lean):
   *   receiverKnowsTrajectory(deficit, deficit) = 0
   */
  static teleportReceive(channel: NeuralMeshTeleportationChannel): NeuralMeshTeleportationResult {
    const layers: LayerConvergenceTrajectory[] = [];

    for (let i = 0; i < channel.numLayers; i++) {
      const deficit = channel.layerDeficits[i];
      if (deficit === undefined) {
        throw new Error('teleportReceive: missing layer deficit');
      }
      const trajectory: number[] = [];
      for (let k = 0; k <= deficit; k++) {
        // teleportation_trajectory_from_deficit: deficit - min(k, deficit)
        trajectory.push(deficit - k);
      }
      layers.push({
        layerIdx: i,
        deficit,
        trajectory,
        convergenceRound: deficit,
        isConverged: deficit === 0,
        // God formula: dampening = 1 - deficit/(9 * rounds)
        // approximation of how much this layer is dampened
        dampeningEstimate: channel.rounds > 0
          ? deficit / (9 * channel.rounds)
          : 0,
      });
    }

    return {
      layers,
      totalDeficit: channel.layerDeficits.reduce((s, d) => s + d, 0),
      allConverged: channel.layerDeficits.every(d => d === 0),
      meshHealth: channel.rounds > 0
        ? 1 - channel.layerDeficits.reduce((s, d) => s + d, 0) / (9 * channel.numLayers * channel.rounds)
        : 1,
    };
  }
}

// ═══════════════════════════════════════════════════════════════════════
// STATISTICAL TELEPORTATION TYPES
// ═══════════════════════════════════════════════════════════════════════
//
// From StatisticalTeleportation.lean:
//   "Certainty transfers. Data stays local."
//
// The teleportation channel carries per-layer deficit integers — one
// natural number per layer that encodes the full future convergence
// trajectory. The receiver reconstructs trajectories without seeing
// any weight data or rejection history.

/**
 * Teleportation channel: the minimal wire payload for distributed
 * NeuralMesh inference. One deficit integer per layer.
 *
 * Structurally identical to quantum teleportation:
 *   - Entangled pair = shared void tunnel (the mesh topology)
 *   - Classical channel = deficit integers crossing the wire
 *   - Teleported state = certainty (convergence trajectory)
 */
export interface NeuralMeshTeleportationChannel {
  /** Per-layer deficit: deficit = totalRejections - minRejection × 9 */
  readonly layerDeficits: readonly number[];
  /** Number of layers in the source mesh */
  readonly numLayers: number;
  /** Beta1 per cell (always 9 for Sophia) */
  readonly beta1: number;
  /** Composite dimension of the source mesh */
  readonly compositeDimension: number;
  /** Total god formula rounds at time of transmission */
  readonly rounds: number;
  /** Timestamp of transmission */
  readonly transmittedAt: number;
}

/**
 * Per-layer convergence trajectory reconstructed from a deficit.
 *
 * teleportation_trajectory_from_deficit:
 *   deficit at round k = max(0, deficit - k)
 *
 * teleportation_convergence_round:
 *   convergence at round = deficit (exactly)
 *
 * teleportation_monotone:
 *   trajectory only decreases
 */
export interface LayerConvergenceTrajectory {
  /** Layer index */
  readonly layerIdx: number;
  /** The deficit integer (the teleported payload) */
  readonly deficit: number;
  /** Full trajectory: deficit at each future round [deficit, deficit-1, ..., 0] */
  readonly trajectory: readonly number[];
  /** Exact convergence round */
  readonly convergenceRound: number;
  /** Whether this layer has already converged (deficit = 0) */
  readonly isConverged: boolean;
  /** Estimated dampening factor (0 = healthy, 1 = fully dampened) */
  readonly dampeningEstimate: number;
}

/**
 * Result of receiving a teleportation channel.
 */
export interface NeuralMeshTeleportationResult {
  /** Per-layer convergence trajectories */
  readonly layers: readonly LayerConvergenceTrajectory[];
  /** Total deficit across all layers */
  readonly totalDeficit: number;
  /** Whether all layers have converged */
  readonly allConverged: boolean;
  /** Overall mesh health (1 = fully healthy, 0 = fully dampened) */
  readonly meshHealth: number;
}

// ═══════════════════════════════════════════════════════════════════════
// FACTORY
// ═══════════════════════════════════════════════════════════════════════

/**
 * Create a NeuralMesh from a model config.
 *
 * @example
 * ```ts
 * const mesh = createNeuralMesh({
 *   hiddenDim: 2048, numHeads: 32, numKvHeads: 4,
 *   headDim: 64, intermediateSize: 5632, numLayers: 22,
 * });
 * mesh.loadAllWeights(weights);
 * const output = mesh.forward(inputHidden, position);
 * ```
 */
export function createNeuralMesh(config: NeuralMeshModelConfig): NeuralMesh {
  return new NeuralMesh(config);
}

/**
 * Create a NeuralMesh auto-scaled to a parameter count.
 * Uses resolveSyzygyDimension to pick the appropriate gnostic level.
 */
export function createNeuralMeshForParameters(
  config: NeuralMeshModelConfig,
): NeuralMesh {
  // The mesh is always Sophia (k=3, s=3, β₁=9) because that's the
  // natural match for transformer weight tensor count.
  // The gnostic level determines the inter-mesh composition dimension
  // when multiple NeuralMeshes compose (future: model ensemble).
  return new NeuralMesh(config);
}

/**
 * Create random weights for testing.
 * Produces normally-distributed weights scaled by 1/√dim.
 */
export function createRandomCellWeights(
  config: NeuralMeshModelConfig,
): CellWeights {
  const { hiddenDim, numHeads, numKvHeads, headDim, intermediateSize } = config;
  const qDim = numHeads * headDim;
  const kvDim = numKvHeads * headDim;

  const randNorm = (size: number, scale: number): Float32Array => {
    const arr = new Float32Array(size);
    for (let i = 0; i < size; i++) {
      const u1 = Math.random();
      const u2 = Math.random();
      arr[i] = Math.sqrt(-2 * Math.log(u1)) * Math.cos(2 * Math.PI * u2) * scale;
    }
    return arr;
  };

  return {
    attnNorm: new Float32Array(hiddenDim).fill(1),
    qProj: randNorm(qDim * hiddenDim, 1 / Math.sqrt(hiddenDim)),
    ffnNorm: new Float32Array(hiddenDim).fill(1),
    kProj: randNorm(kvDim * hiddenDim, 1 / Math.sqrt(hiddenDim)),
    vProj: randNorm(kvDim * hiddenDim, 1 / Math.sqrt(hiddenDim)),
    gateProj: randNorm(intermediateSize * hiddenDim, 1 / Math.sqrt(hiddenDim)),
    oProj: randNorm(hiddenDim * qDim, 1 / Math.sqrt(qDim)),
    upProj: randNorm(intermediateSize * hiddenDim, 1 / Math.sqrt(hiddenDim)),
    downProj: randNorm(hiddenDim * intermediateSize, 1 / Math.sqrt(intermediateSize)),
  };
}
