// @a0n/.ignored_gnosis/src/organism-topology.ts
/**
 * Organism Topology — Per-project health from real code topology
 *
 * Walks source files in a project, runs the polyglot scanner on each,
 * and maps the results through the polyglot-checkpoint bridge to produce
 * a CheckpointedSystem with real sigma — not estimates from rg -c, but
 * actual beta1 and deficit from tree-sitter ASTs fed through Betty.
 *
 * Each project becomes an OrganismNode in the workspace topology.
 * The organism formalizes the codebase.
 *
 * @module @a0n/gnosis/organism-topology
 */

import { readdirSync, readFileSync, statSync } from 'fs';
import { extname, join, relative, resolve } from 'path';
import { analyzePolyglotSource } from './polyglot-bridge';
import type { PolyglotAnalysisResult } from './polyglot-bridge';
import {
  type ProjectCheckpointProfile,
  type ProjectScanResult,
  diagnoseProject,
  diagnoseFromCounts,
  aggregateFindings,
} from './polyglot-checkpoint-bridge';
import type { HealthStatus } from './checkpoint-substrates';
import {
  loadCachedScan,
  saveScanToCache,
} from './polyglot-scan-cache';

// ═══════════════════════════════════════════════════════════════════════
// Types
// ═══════════════════════════════════════════════════════════════════════

/**
 * A single project in the workspace organism.
 */
export interface OrganismNode {
  /** Project name (directory name or package name). */
  name: string;
  /** Absolute path to project root. */
  root: string;
  /** Relative path from workspace root. */
  relativePath: string;
  /** Source file count. */
  fileCount: number;
  /** Real sigma from checkpoint homology (sicknessFraction). */
  sigma: number;
  /** Health classification from checkpoint algebra. */
  health: HealthStatus;
  /** Full checkpoint profile (null if scan failed). */
  profile: ProjectCheckpointProfile | null;
  /** Errors encountered during scanning. */
  errors: string[];
}

/**
 * Full organism topology for a workspace.
 */
export interface OrganismTopology {
  /** Workspace root. */
  workspaceRoot: string;
  /** Per-project nodes. */
  nodes: OrganismNode[];
  /** Aggregate stats. */
  stats: OrganismStats;
}

export interface OrganismStats {
  /** Total projects scanned. */
  projectCount: number;
  /** Projects with health = 'healthy'. */
  healthyCount: number;
  /** Projects with health = 'stressed' or 'inflamed'. */
  stressedCount: number;
  /** Projects with health = 'cancerous'. */
  cancerousCount: number;
  /** Projects with health = 'necrotic'. */
  necroticCount: number;
  /** Average sigma across all projects. */
  avgSigma: number;
  /** Maximum sigma (worst project). */
  maxSigma: number;
  /** Total functions analyzed. */
  totalFunctions: number;
  /** Total hotspots (godWeight > threshold). */
  totalHotspots: number;
}

// ═══════════════════════════════════════════════════════════════════════
// File Discovery
// ═══════════════════════════════════════════════════════════════════════

const IGNORED_DIRS = new Set([
  'node_modules',
  '.next',
  'dist',
  'build',
  '.git',
  'coverage',
  '.wrangler',
  '__tests__',
  '.a0',
  '.lake',
  '.tmp',
  'storybook-static',
  'target',
  '.turbo',
]);

const SCANNABLE_EXTENSIONS = new Set([
  '.ts',
  '.tsx',
  '.js',
  '.jsx',
  '.py',
  '.rs',
  '.go',
  '.rb',
  '.java',
  '.kt',
  '.swift',
  '.c',
  '.cpp',
  '.h',
  '.hpp',
]);

function walkSourceFiles(root: string): string[] {
  const files: string[] = [];
  const queue = [root];

  while (queue.length > 0) {
    const dir = queue.pop()!;
    let entries;
    try {
      entries = readdirSync(dir, { withFileTypes: true });
    } catch {
      continue;
    }

    for (const entry of entries) {
      if (entry.isDirectory()) {
        if (!IGNORED_DIRS.has(entry.name)) queue.push(join(dir, entry.name));
      } else if (SCANNABLE_EXTENSIONS.has(extname(entry.name))) {
        files.push(join(dir, entry.name));
      }
    }
  }

  return files;
}

// ═══════════════════════════════════════════════════════════════════════
// Project Scanning
// ═══════════════════════════════════════════════════════════════════════

/**
 * Scan a single project and produce its checkpoint profile.
 *
 * Walks all source files, runs polyglot analysis on each, and maps
 * through the bridge to a CheckpointedSystem with real sigma.
 */
export async function scanProject(
  projectName: string,
  projectRoot: string,
): Promise<OrganismNode & { profile: ProjectCheckpointProfile }> {
  const files = walkSourceFiles(projectRoot);
  const fileResults: Array<{
    filePath: string;
    result: PolyglotAnalysisResult;
  }> = [];
  const errors: string[] = [];

  for (const filePath of files) {
    try {
      const result = await analyzePolyglotSource(filePath);
      fileResults.push({ filePath, result });
    } catch (e: unknown) {
      const msg = e instanceof Error ? e.message : String(e);
      errors.push(`${filePath}: ${msg}`);
    }
  }

  const profile = diagnoseProject(projectName, fileResults);

  return {
    name: projectName,
    root: projectRoot,
    relativePath: projectRoot, // caller should fix up
    fileCount: files.length,
    sigma: profile.sigma,
    health: profile.diagnosis.health,
    profile,
    errors: [...errors, ...profile.scan.errors],
  };
}

/**
 * Scan a project without the polyglot binary — uses file/function counting
 * and structural heuristics as a lightweight estimator.
 *
 * A function is classified as "active" (has a feedback channel) when ANY of:
 *   1. It's in a test file (direct test coverage)
 *   2. It's in a TypeScript file (type checker = feedback channel)
 *   3. It has control flow loops (beta1 > 0 proxy — can learn)
 *   4. It has error handling (try/catch = explicit feedback path)
 *
 * A function is "destroyed" (deaf) only when it has NONE of these.
 * This avoids the false-positive flood where all non-test TS functions
 * are marked deaf despite having type-checking feedback.
 *
 * The real scanner (polyglot binary → Betty) produces exact numbers.
 * This lightweight path is for when the Rust binary isn't available.
 */
export function scanProjectLightweight(
  projectName: string,
  projectRoot: string,
): OrganismNode {
  const files = walkSourceFiles(projectRoot);
  let activeFunctions = 0;
  let destroyedFunctions = 0;

  for (const filePath of files) {
    try {
      const content = readFileSync(filePath, 'utf8');
      // Rough function count: function declarations + arrow functions
      const funcMatches = content.match(
        /\bfunction\s+\w+|(?:const|let|var)\s+\w+\s*=\s*(?:async\s*)?\(/g,
      );
      const count = funcMatches?.length ?? 0;
      if (count === 0) continue;

      const isTest =
        filePath.includes('.test.') ||
        filePath.includes('.spec.') ||
        filePath.includes('__tests__');
      const isTyped =
        filePath.endsWith('.ts') || filePath.endsWith('.tsx');
      const hasLoops = /\b(for|while|do)\s*\(/.test(content);
      const hasErrorHandling = /\btry\s*\{/.test(content);

      // A function has a feedback channel if ANY signal is present
      if (isTest || isTyped || hasLoops || hasErrorHandling) {
        activeFunctions += count;
      } else {
        destroyedFunctions += count;
      }
    } catch {
      // Skip unreadable files
    }
  }

  const quick = diagnoseFromCounts(projectName, activeFunctions, destroyedFunctions);

  return {
    name: projectName,
    root: projectRoot,
    relativePath: projectRoot,
    fileCount: files.length,
    sigma: quick.sigma,
    health: quick.diagnosis.health,
    profile: null,
    errors: [],
  };
}

// ═══════════════════════════════════════════════════════════════════════
// Workspace Topology
// ═══════════════════════════════════════════════════════════════════════

/**
 * Discover project directories in a workspace.
 * Looks for directories containing package.json, Cargo.toml, go.mod, etc.
 */
export function discoverProjects(workspaceRoot: string): string[] {
  const markers = [
    'package.json',
    'Cargo.toml',
    'go.mod',
    'pyproject.toml',
    'setup.py',
    'Gemfile',
    'build.gradle',
    'pom.xml',
  ];

  const projects: string[] = [];
  const queue = [workspaceRoot];
  const visited = new Set<string>();

  while (queue.length > 0) {
    const dir = queue.pop()!;
    if (visited.has(dir)) continue;
    visited.add(dir);

    let entries;
    try {
      entries = readdirSync(dir, { withFileTypes: true });
    } catch {
      continue;
    }

    const hasMarker = entries.some(
      (e) => !e.isDirectory() && markers.includes(e.name),
    );
    if (hasMarker && dir !== workspaceRoot) {
      projects.push(dir);
      // Don't descend into sub-projects of a project
      continue;
    }

    for (const entry of entries) {
      if (entry.isDirectory() && !IGNORED_DIRS.has(entry.name)) {
        queue.push(join(dir, entry.name));
      }
    }
  }

  return projects;
}

/**
 * Compute the full organism topology for a workspace.
 *
 * Uses cached scan results when available (invalidated by content hash).
 * Pass `full: true` to bypass cache and re-scan everything.
 * Falls back to lightweight scanning if the scanner fails for a project.
 */
export async function computeOrganismTopology(
  workspaceRoot: string,
  projectDirs?: string[],
  options?: { full?: boolean },
): Promise<OrganismTopology> {
  const dirs = projectDirs ?? discoverProjects(workspaceRoot);
  const useCache = !options?.full;

  const nodes: OrganismNode[] = [];

  for (const dir of dirs) {
    const name = relative(workspaceRoot, dir) || dir;

    // Try cache first (unless --full)
    if (useCache) {
      const cached = loadCachedScan(workspaceRoot, name, dir);
      if (cached) {
        // Rebuild profile from cached scan
        const profile = diagnoseProject(name, cached.functions.map((f) => ({
          filePath: f.filePath,
          result: {
            filePath: f.filePath,
            language: 'cached',
            structuredDeclarations: [],
            proofDeclarations: [],
            functions: [],
            errors: [],
          },
        })));
        // Use the cached scan's counts directly
        const quick = diagnoseFromCounts(
          name,
          cached.functions.filter((f) => f.isLearning).length,
          cached.functions.filter((f) => !f.isLearning).length,
        );
        nodes.push({
          name,
          root: dir,
          relativePath: relative(workspaceRoot, dir),
          fileCount: cached.filesAnalyzed,
          sigma: quick.sigma,
          health: quick.diagnosis.health,
          profile: null, // cached — no full profile
          errors: [],
        });
        continue;
      }
    }

    try {
      const node = await scanProject(name, dir);
      node.relativePath = relative(workspaceRoot, dir);
      // Cache the scan result
      if (node.profile) {
        saveScanToCache(workspaceRoot, name, dir, node.profile.scan);
      }
      nodes.push(node);
    } catch {
      // Fall back to lightweight scan
      const node = scanProjectLightweight(name, dir);
      node.relativePath = relative(workspaceRoot, dir);
      nodes.push(node);
    }
  }

  const stats = computeStats(nodes);

  return { workspaceRoot, nodes, stats };
}

/**
 * Lightweight organism topology — no polyglot binary needed.
 */
export function computeOrganismTopologyLightweight(
  workspaceRoot: string,
  projectDirs?: string[],
): OrganismTopology {
  const dirs = projectDirs ?? discoverProjects(workspaceRoot);

  const nodes: OrganismNode[] = dirs.map((dir) => {
    const name = relative(workspaceRoot, dir) || dir;
    const node = scanProjectLightweight(name, dir);
    node.relativePath = relative(workspaceRoot, dir);
    return node;
  });

  const stats = computeStats(nodes);

  return { workspaceRoot, nodes, stats };
}

function computeStats(nodes: OrganismNode[]): OrganismStats {
  const sigmas = nodes.map((n) => n.sigma);
  const totalSigma = sigmas.reduce((s, v) => s + v, 0);

  return {
    projectCount: nodes.length,
    healthyCount: nodes.filter((n) => n.health === 'healthy').length,
    stressedCount: nodes.filter(
      (n) => n.health === 'stressed' || n.health === 'inflamed',
    ).length,
    cancerousCount: nodes.filter((n) => n.health === 'cancerous').length,
    necroticCount: nodes.filter((n) => n.health === 'necrotic').length,
    avgSigma: nodes.length > 0 ? totalSigma / nodes.length : 0,
    maxSigma: sigmas.length > 0 ? Math.max(...sigmas) : 0,
    totalFunctions: nodes.reduce(
      (s, n) => s + (n.profile?.scan.functionCount ?? 0),
      0,
    ),
    totalHotspots: nodes.reduce(
      (s, n) => s + (n.profile?.scan.hotspotCount ?? 0),
      0,
    ),
  };
}

// ═══════════════════════════════════════════════════════════════════════
// Reporting
// ═══════════════════════════════════════════════════════════════════════

const HEALTH_SYMBOLS: Record<HealthStatus, string> = {
  healthy: 'ok',
  stressed: '!!',
  inflamed: '**',
  cancerous: 'XX',
  necrotic: '--',
};

/**
 * Format organism topology as a text report.
 */
export function formatOrganismReport(topology: OrganismTopology): string {
  const lines: string[] = [];
  const { stats } = topology;

  lines.push(`Organism Topology: ${topology.workspaceRoot}`);
  lines.push(`${'='.repeat(60)}`);
  lines.push(
    `Projects: ${stats.projectCount} | ` +
      `Healthy: ${stats.healthyCount} | ` +
      `Stressed: ${stats.stressedCount} | ` +
      `Cancerous: ${stats.cancerousCount} | ` +
      `Necrotic: ${stats.necroticCount}`,
  );
  lines.push(
    `Avg sigma: ${stats.avgSigma.toFixed(3)} | ` +
      `Max sigma: ${stats.maxSigma.toFixed(3)} | ` +
      `Functions: ${stats.totalFunctions} | ` +
      `Hotspots: ${stats.totalHotspots}`,
  );
  lines.push('');

  // Sort by sigma descending (worst first)
  const sorted = [...topology.nodes].sort((a, b) => b.sigma - a.sigma);

  for (const node of sorted) {
    const sym = HEALTH_SYMBOLS[node.health];
    const sigma = node.sigma.toFixed(3);
    const files = `${node.fileCount} files`;
    const funcs = node.profile
      ? `${node.profile.scan.functionCount} funcs`
      : 'lightweight';
    const repair = node.profile
      ? `repair: ${node.profile.diagnosis.repairPath}`
      : '';

    lines.push(
      `  [${sym}] ${node.relativePath.padEnd(40)} sigma=${sigma}  ${files}  ${funcs}  ${repair}`,
    );

    // Show top repair candidates if profile available
    if (node.profile && node.profile.repairOrdering.length > 0) {
      const top3 = node.profile.repairOrdering.slice(0, 3);
      for (const candidate of top3) {
        lines.push(
          `        -> ${candidate.functionName} (godWeight=${candidate.godWeight}, beta1=${candidate.beta1})`,
        );
      }
    }
  }

  return lines.join('\n');
}
