// @a0n/.ignored_gnosis/src/phyle-comic-parallax.ts
/**
 * phyle-comic-parallax.ts — Depth-aware comic panel parallax from Phyle temporal depth.
 *
 * Takes the fused depth + standing wave mask from the temporal pipeline and
 * produces layered parallax data for animated comic viewing:
 *
 * 1. Layer separation: foreground/midground/background via depth quantization
 * 2. Text slot ranking: most stable (standing wave) regions get primary text
 * 3. Parallax motion vectors: per-layer displacement for 3D panel effect
 * 4. Edge-preserved layer masks: deblur-stable edges define layer boundaries
 *
 * The artistic vision: static comic panels become living dioramas.
 * Characters float in front of backgrounds. Speech bubbles anchor to
 * standing-wave regions (confirmed stable across the temporal triton).
 * Noise (compression artifacts, scan noise) is filtered out rather than
 * rendered as phantom depth.
 */

import type { PhyleTemporalPipelineState } from './phyle-temporal-depth';

// ─── Types ──────────────────────────────────────────────────────────

export interface ParallaxLayer {
  id: 'foreground' | 'midground' | 'background';
  depthRange: [number, number];
  pixelMask: Float32Array;
  pixelCount: number;
  meanDepth: number;
  parallaxScale: number;
  standingWaveCoverage: number;
}

export interface TextSlot {
  x0: number;
  y0: number;
  x1: number;
  y1: number;
  area: number;
  layer: 'foreground' | 'midground' | 'background';
  standingWaveConfidence: number;
  deblurStability: number;
}

export interface ComicParallaxResult {
  width: number;
  height: number;
  layers: [ParallaxLayer, ParallaxLayer, ParallaxLayer];
  textSlots: TextSlot[];
  depthMap: Float32Array;
  edgeMask: Float32Array;
  noiseRejected: number;
}

// ─── Layer Separation ───────────────────────────────────────────────

const FOREGROUND_THRESHOLD = 0.33;
const MIDGROUND_THRESHOLD = 0.66;

/**
 * Handles the gnosis separate Parallax Layers workflow.
 */
export function separateParallaxLayers(
  state: PhyleTemporalPipelineState,
): ComicParallaxResult | null {
  if (!state.fusedDepth || !state.standingWaveMask || !state.deblurStability) {
    return null;
  }

  const { width, height } = state;
  const pix = width * height;

  // Normalize fused depth to [0,1]
  let minDepth = Infinity;
  let maxDepth = -Infinity;
  for (let i = 0; i < pix; i++) {
    const d = state.fusedDepth[i]!;
    if (d > 0) {
      minDepth = Math.min(minDepth, d);
      maxDepth = Math.max(maxDepth, d);
    }
  }
  if (minDepth >= maxDepth) {
    minDepth = 0;
    maxDepth = 1;
  }
  const range = maxDepth - minDepth;

  const normalizedDepth = new Float32Array(pix);
  for (let i = 0; i < pix; i++) {
    normalizedDepth[i] = (state.fusedDepth[i]! - minDepth) / range;
  }

  // Separate into 3 layers
  const fgMask = new Float32Array(pix);
  const mgMask = new Float32Array(pix);
  const bgMask = new Float32Array(pix);

  let fgCount = 0, mgCount = 0, bgCount = 0;
  let fgDepthSum = 0, mgDepthSum = 0, bgDepthSum = 0;
  let fgSwSum = 0, mgSwSum = 0, bgSwSum = 0;
  let noiseRejected = 0;

  for (let i = 0; i < pix; i++) {
    const d = normalizedDepth[i]!;
    const sw = state.standingWaveMask[i]!;

    // Reject pixels with no standing wave confirmation (white noise)
    if (sw === 0) {
      noiseRejected++;
      continue;
    }

    if (d < FOREGROUND_THRESHOLD) {
      fgMask[i] = sw;
      fgCount++;
      fgDepthSum += d;
      fgSwSum += sw;
    } else if (d < MIDGROUND_THRESHOLD) {
      mgMask[i] = sw;
      mgCount++;
      mgDepthSum += d;
      mgSwSum += sw;
    } else {
      bgMask[i] = sw;
      bgCount++;
      bgDepthSum += d;
      bgSwSum += sw;
    }
  }

  const layers: [ParallaxLayer, ParallaxLayer, ParallaxLayer] = [
    {
      id: 'foreground',
      depthRange: [0, FOREGROUND_THRESHOLD],
      pixelMask: fgMask,
      pixelCount: fgCount,
      meanDepth: fgCount > 0 ? fgDepthSum / fgCount : 0,
      parallaxScale: 1.0,
      standingWaveCoverage: fgCount > 0 ? fgSwSum / fgCount : 0,
    },
    {
      id: 'midground',
      depthRange: [FOREGROUND_THRESHOLD, MIDGROUND_THRESHOLD],
      pixelMask: mgMask,
      pixelCount: mgCount,
      meanDepth: mgCount > 0 ? mgDepthSum / mgCount : 0,
      parallaxScale: 0.5,
      standingWaveCoverage: mgCount > 0 ? mgSwSum / mgCount : 0,
    },
    {
      id: 'background',
      depthRange: [MIDGROUND_THRESHOLD, 1.0],
      pixelMask: bgMask,
      pixelCount: bgCount,
      meanDepth: bgCount > 0 ? bgDepthSum / bgCount : 0,
      parallaxScale: 0.2,
      standingWaveCoverage: bgCount > 0 ? bgSwSum / bgCount : 0,
    },
  ];

  // Build edge mask from deblur stability
  const edgeMask = new Float32Array(pix);
  for (let i = 0; i < pix; i++) {
    edgeMask[i] = state.deblurStability[i]! > 0.8 ? 1.0 : 0.0;
  }

  return {
    width,
    height,
    layers,
    textSlots: [],
    depthMap: normalizedDepth,
    edgeMask,
    noiseRejected,
  };
}

// ─── Text Slot Ranking ──────────────────────────────────────────────

/**
 * Handles the gnosis rank Text Slots workflow.
 */
export function rankTextSlots(
  result: ComicParallaxResult,
  textRegionMask: Uint8Array,
): TextSlot[] {
  const { width, height } = result;
  const pix = width * height;
  const visited = new Uint8Array(pix);
  const slots: TextSlot[] = [];

  // Find connected components in text region mask
  for (let start = 0; start < pix; start++) {
    if (textRegionMask[start]! < 128 || visited[start]) continue;

    // BFS flood fill
    const queue: number[] = [start];
    visited[start] = 1;
    let minX = start % width, maxX = minX;
    let minY = Math.floor(start / width), maxY = minY;
    let swSum = 0, dsSum = 0, depthSum = 0, count = 0;

    while (queue.length > 0) {
      const pos = queue.shift()!;
      const x = pos % width;
      const y = Math.floor(pos / width);
      minX = Math.min(minX, x);
      maxX = Math.max(maxX, x);
      minY = Math.min(minY, y);
      maxY = Math.max(maxY, y);
      swSum += result.layers[0].pixelMask[pos]! + result.layers[1].pixelMask[pos]! + result.layers[2].pixelMask[pos]!;
      dsSum += result.edgeMask[pos]! > 0 ? 0 : 1; // interior (non-edge) = good for text
      depthSum += result.depthMap[pos]!;
      count++;

      for (const [dx, dy] of [[0, -1], [0, 1], [-1, 0], [1, 0]] as const) {
        const nx = x + dx, ny = y + dy;
        if (nx >= 0 && nx < width && ny >= 0 && ny < height) {
          const ni = ny * width + nx;
          if (textRegionMask[ni]! >= 128 && !visited[ni]) {
            visited[ni] = 1;
            queue.push(ni);
          }
        }
      }
    }

    // Filter small regions
    const area = (maxX - minX + 1) * (maxY - minY + 1);
    if (area < 100) continue;

    const meanDepth = count > 0 ? depthSum / count : 0.5;
    const layer: TextSlot['layer'] =
      meanDepth < FOREGROUND_THRESHOLD ? 'foreground' :
      meanDepth < MIDGROUND_THRESHOLD ? 'midground' : 'background';

    slots.push({
      x0: minX,
      y0: minY,
      x1: maxX + 1,
      y1: maxY + 1,
      area,
      layer,
      standingWaveConfidence: count > 0 ? swSum / count : 0,
      deblurStability: count > 0 ? dsSum / count : 0,
    });
  }

  // Sort: highest standing wave confidence first (most stable = best for text)
  slots.sort((a, b) => b.standingWaveConfidence - a.standingWaveConfidence);
  result.textSlots = slots;
  return slots;
}

// ─── Parallax Motion Vectors ────────────────────────────────────────

export interface ParallaxMotion {
  layerId: 'foreground' | 'midground' | 'background';
  dx: number;
  dy: number;
}

/**
 * Computes the Parallax Motion.
 */
export function computeParallaxMotion(
  result: ComicParallaxResult,
  viewAngleX: number,
  viewAngleY: number,
): ParallaxMotion[] {
  return result.layers.map(layer => ({
    layerId: layer.id,
    dx: viewAngleX * layer.parallaxScale * 20,
    dy: viewAngleY * layer.parallaxScale * 20,
  }));
}

// ─── Export for animated panel rendering ────────────────────────────

export interface AnimatedPanelFrame {
  layers: Array<{
    id: string;
    offsetX: number;
    offsetY: number;
    opacity: number;
  }>;
  timestamp: number;
}

/**
 * Handles the gnosis generate Animated Panel Frames workflow.
 */
export function generateAnimatedPanelFrames(
  result: ComicParallaxResult,
  durationMs: number,
  fps: number,
): AnimatedPanelFrame[] {
  const frameCount = Math.ceil(durationMs / 1000 * fps);
  const frames: AnimatedPanelFrame[] = [];

  for (let i = 0; i < frameCount; i++) {
    const t = i / frameCount;
    const angle = t * Math.PI * 2;

    // Gentle circular parallax motion
    const viewX = Math.sin(angle) * 0.3;
    const viewY = Math.cos(angle) * 0.15;
    const motions = computeParallaxMotion(result, viewX, viewY);

    frames.push({
      layers: motions.map(m => ({
        id: m.layerId,
        offsetX: m.dx,
        offsetY: m.dy,
        opacity: 1.0,
      })),
      timestamp: (i / fps) * 1000,
    });
  }

  return frames;
}
