// @a0n/.ignored_gnosis/src/polyglot-import.ts
/**
 * Topology imports -- composable verified modules.
 *
 * Resolves @import directives in .gg files, merges imported topologies
 * into the host topology, and deduplicates node IDs.
 *
 * Syntax:
 *   @import "auth_flow.gg"
 *   @import "./lib/data_pipeline.gg"
 *   @import "@affectively/common-topologies/rate-limiter.gg"
 *
 * The imported topology's nodes are prefixed with the import name to
 * avoid collisions. Edges that cross import boundaries use the prefixed IDs.
 */

import * as fs from 'node:fs';
import * as path from 'node:path';
import {
  BettyCompiler,
  type GraphAST,
  type ASTNode,
  type ASTEdge,
} from './betty/compiler';

/**
 * Resolved import directive.
 */
export interface ResolvedImport {
  /** The import path as written in the .gg source. */
  importPath: string;
  /** The resolved absolute file path. */
  resolvedPath: string;
  /** The prefix applied to imported node IDs. */
  prefix: string;
  /** The parsed AST of the imported topology. */
  ast: GraphAST;
  /** Any nested imports resolved recursively. */
  nestedImports: ResolvedImport[];
}

/**
 * Result of import resolution.
 */
export interface ImportResolutionResult {
  /** The merged AST with all imports inlined. */
  mergedAst: GraphAST;
  /** The merged .gg source (for debugging). */
  mergedSource: string;
  /** All resolved imports (flat list). */
  imports: ResolvedImport[];
  /** Errors encountered during resolution. */
  errors: string[];
}

/**
 * Parse @import directives from .gg source text.
 */
function parseImports(source: string): string[] {
  const imports: string[] = [];
  const importPattern = /^@import\s+"([^"]+)"/gm;
  let match: RegExpExecArray | null;
  while ((match = importPattern.exec(source)) !== null) {
    imports.push(match[1]!);
  }
  return imports;
}

/**
 * Remove @import lines from source (they are directives, not GG syntax).
 */
function stripImports(source: string): string {
  return source.replace(/^@import\s+"[^"]+"\s*$/gm, '').trim();
}

/**
 * Generate a prefix from an import path.
 * "./lib/auth_flow.gg" → "auth_flow"
 * "@affectively/rate-limiter.gg" → "rate_limiter"
 */
function importPrefix(importPath: string): string {
  return path
    .basename(importPath, '.gg')
    .replace(/[^a-zA-Z0-9]/g, '_')
    .toLowerCase();
}

/**
 * Prefix all node IDs and edge references in an AST.
 */
function prefixAst(ast: GraphAST, prefix: string): GraphAST {
  const newNodes = new Map<string, ASTNode>();
  for (const [id, node] of ast.nodes) {
    const newId = `${prefix}_${id}`;
    newNodes.set(newId, {
      ...node,
      id: newId,
    });
  }

  const newEdges: ASTEdge[] = ast.edges.map((edge) => ({
    ...edge,
    sourceIds: edge.sourceIds.map((id) => `${prefix}_${id.trim()}`),
    targetIds: edge.targetIds.map((id) => `${prefix}_${id.trim()}`),
  }));

  return { nodes: newNodes, edges: newEdges };
}

/**
 * Merge two ASTs into one (additive -- no deduplication of nodes).
 */
function mergeAsts(base: GraphAST, addition: GraphAST): GraphAST {
  const nodes = new Map(base.nodes);
  for (const [id, node] of addition.nodes) {
    nodes.set(id, node);
  }
  const edges = [...base.edges, ...addition.edges];
  return { nodes, edges };
}

/**
 * Resolve all @import directives in a .gg file, recursively.
 *
 * @param source The .gg source text.
 * @param basePath The directory of the .gg file (for relative import resolution).
 * @param visited Set of already-resolved paths (cycle detection).
 */
export function resolveImports(
  source: string,
  basePath: string,
  visited: Set<string> = new Set()
): ImportResolutionResult {
  const errors: string[] = [];
  const allImports: ResolvedImport[] = [];
  const importPaths = parseImports(source);
  const strippedSource = stripImports(source);

  // Parse the base topology.
  const compiler = new BettyCompiler();
  const baseResult = compiler.parse(strippedSource);

  if (!baseResult.ast) {
    return {
      mergedAst: { nodes: new Map(), edges: [] },
      mergedSource: strippedSource,
      imports: [],
      errors: [
        `Failed to parse base topology`,
        ...baseResult.diagnostics.map(
          (d) => `${d.line}:${d.column} ${d.message}`
        ),
      ],
    };
  }

  let mergedAst = baseResult.ast;
  let mergedSource = strippedSource;

  // Resolve each import.
  for (const importPath of importPaths) {
    // Resolve file path.
    let resolvedPath: string;
    if (importPath.startsWith('.') || importPath.startsWith('/')) {
      resolvedPath = path.resolve(basePath, importPath);
    } else if (importPath.startsWith('@')) {
      // Package import -- look in node_modules or a topology registry.
      resolvedPath = path.resolve(basePath, 'node_modules', importPath);
    } else {
      resolvedPath = path.resolve(basePath, importPath);
    }

    // Cycle detection.
    if (visited.has(resolvedPath)) {
      errors.push(`Circular import detected: ${importPath} (${resolvedPath})`);
      continue;
    }
    visited.add(resolvedPath);

    // Read and parse the imported file.
    if (!fs.existsSync(resolvedPath)) {
      errors.push(
        `Import not found: ${importPath} (resolved to ${resolvedPath})`
      );
      continue;
    }

    const importSource = fs.readFileSync(resolvedPath, 'utf8');
    const prefix = importPrefix(importPath);

    // Recursively resolve nested imports.
    const nested = resolveImports(
      importSource,
      path.dirname(resolvedPath),
      new Set(visited)
    );

    if (nested.errors.length > 0) {
      errors.push(...nested.errors.map((e) => `[${importPath}] ${e}`));
    }

    // Prefix the imported AST.
    const prefixedAst = prefixAst(nested.mergedAst, prefix);

    // Merge into the host AST.
    mergedAst = mergeAsts(mergedAst, prefixedAst);
    mergedSource += `\n\n// @import "${importPath}" (prefix: ${prefix})\n${nested.mergedSource}`;

    allImports.push({
      importPath,
      resolvedPath,
      prefix,
      ast: prefixedAst,
      nestedImports: nested.imports,
    });
    allImports.push(...nested.imports);
  }

  return {
    mergedAst,
    mergedSource,
    imports: allImports,
    errors,
  };
}
