// @a0n/.ignored_gnosis/src/polyglot-vfs.ts
/**
 * Polyglot Virtual File System (VFS).
 *
 * Pre-parses all source files in a repository into GG topology fragments,
 * stores them in D1 (Cloudflare Workers), and serves them with sub-1ms
 * latency at the edge. Any system can load pre-parsed topologies without
 * re-running tree-sitter or the Rust polyglot binary.
 *
 * This means: run any language on Cloudflare Workers. Not by embedding
 * interpreters -- by pre-compiling everything to GG topology fragments
 * stored in D1, then executing through the GnosisEngine with polyglot
 * bridge handlers that dispatch to edge-hosted harnesses.
 *
 * Pipeline:
 *   1. Scan repo: tree-sitter → CFG → GG topology per function
 *   2. Store fragments in D1: key = hash(file+function), value = serialized AST + manifest
 *   3. At edge: load fragment from D1 (sub-1ms), execute via GnosisEngine
 *   4. Harness execution: WASM for simple functions, subprocess for complex ones
 */

import * as fs from 'node:fs';
import * as path from 'node:path';
import { alchemicHash48HexSync } from '@a0n/bitwise/cache-fp64';
import type { GraphAST, ASTNode, ASTEdge } from './betty/compiler';
import type { PolyglotExecutionManifest } from './polyglot-execution-bridge';

/**
 * A pre-parsed topology fragment stored in the VFS.
 */
export interface TopologyFragment {
  /** Unique key: hash(filePath + functionName). */
  key: string;
  /** Source file path (relative to repo root). */
  filePath: string;
  /** Function name. */
  functionName: string;
  /** Detected language. */
  language: string;
  /** Serialized GraphAST (JSON). */
  astJson: string;
  /** Serialized execution manifest (JSON). */
  manifestJson: string;
  /** The .gg source text (for debugging). */
  ggSource: string;
  /** Function signature (JSON). */
  signatureJson: string;
  /** File content hash (for cache invalidation). */
  contentHash: string;
  /** Timestamp of last parse. */
  parsedAt: string;
}

/**
 * D1-compatible database interface (matches x-gnosis's D1LikeDatabase).
 */
export interface D1Database {
  prepare(query: string): D1PreparedStatement;
  batch<T>(statements: D1PreparedStatement[]): Promise<T[]>;
  exec(query: string): Promise<void>;
}

export interface D1PreparedStatement {
  bind(...values: unknown[]): D1PreparedStatement;
  first<T>(): Promise<T | null>;
  all<T>(): Promise<{ results: T[] }>;
  run(): Promise<void>;
}

/**
 * Create the D1 schema for the polyglot VFS.
 */
export function vfsSchema(): string {
  return `
    CREATE TABLE IF NOT EXISTS polyglot_fragments (
      key TEXT PRIMARY KEY,
      file_path TEXT NOT NULL,
      function_name TEXT NOT NULL,
      language TEXT NOT NULL,
      ast_json TEXT NOT NULL,
      manifest_json TEXT NOT NULL,
      gg_source TEXT NOT NULL,
      signature_json TEXT NOT NULL,
      content_hash TEXT NOT NULL,
      parsed_at TEXT NOT NULL
    );

    CREATE INDEX IF NOT EXISTS idx_fragments_file ON polyglot_fragments(file_path);
    CREATE INDEX IF NOT EXISTS idx_fragments_language ON polyglot_fragments(language);
    CREATE INDEX IF NOT EXISTS idx_fragments_hash ON polyglot_fragments(content_hash);
  `;
}

const polyglotFingerprintEncoder = new TextEncoder();

/**
 * Compute a fragment key from file path and function name.
 * **XXH64 fp64** (16 lowercase hex chars), same family as ESC / x-gnosis journal digests.
 * Replaces prior truncated SHA-256 keys — **re-seed D1** `polyglot_fragments` when migrating.
 */
export function fragmentKey(filePath: string, functionName: string): string {
  const a = polyglotFingerprintEncoder.encode(filePath);
  const b = polyglotFingerprintEncoder.encode(functionName);
  const buf = new Uint8Array(a.length + 1 + b.length);
  buf.set(a, 0);
  buf[a.length] = 0;
  buf.set(b, a.length + 1);
  return alchemicHash48HexSync(buf);
}

/**
 * Compute a content fingerprint for cache invalidation (non-cryptographic).
 */
export function contentHash(source: string): string {
  return alchemicHash48HexSync(polyglotFingerprintEncoder.encode(source));
}

/**
 * Store a topology fragment in D1.
 */
export async function storeFragment(
  db: D1Database,
  fragment: TopologyFragment
): Promise<void> {
  await db
    .prepare(
      `INSERT OR REPLACE INTO polyglot_fragments
       (key, file_path, function_name, language, ast_json, manifest_json, gg_source, signature_json, content_hash, parsed_at)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`
    )
    .bind(
      fragment.key,
      fragment.filePath,
      fragment.functionName,
      fragment.language,
      fragment.astJson,
      fragment.manifestJson,
      fragment.ggSource,
      fragment.signatureJson,
      fragment.contentHash,
      fragment.parsedAt
    )
    .run();
}

/**
 * Load a topology fragment from D1 by key.
 * Sub-1ms at edge when D1 is co-located with the Worker.
 */
export async function loadFragment(
  db: D1Database,
  key: string
): Promise<TopologyFragment | null> {
  return db
    .prepare('SELECT * FROM polyglot_fragments WHERE key = ?')
    .bind(key)
    .first<TopologyFragment>();
}

/**
 * Load all fragments for a file.
 */
export async function loadFragmentsByFile(
  db: D1Database,
  filePath: string
): Promise<TopologyFragment[]> {
  const result = await db
    .prepare('SELECT * FROM polyglot_fragments WHERE file_path = ?')
    .bind(filePath)
    .all<TopologyFragment>();
  return result.results;
}

/**
 * Load all fragments for a language.
 */
export async function loadFragmentsByLanguage(
  db: D1Database,
  language: string
): Promise<TopologyFragment[]> {
  const result = await db
    .prepare('SELECT * FROM polyglot_fragments WHERE language = ?')
    .bind(language)
    .all<TopologyFragment>();
  return result.results;
}

/**
 * Check if a fragment is stale (content has changed since last parse).
 */
export async function isFragmentStale(
  db: D1Database,
  filePath: string,
  functionName: string,
  currentContentHash: string
): Promise<boolean> {
  const key = fragmentKey(filePath, functionName);
  const existing = await loadFragment(db, key);
  if (!existing) return true;
  return existing.contentHash !== currentContentHash;
}

/**
 * Deserialize a stored fragment back into executable form.
 */
export function deserializeFragment(fragment: TopologyFragment): {
  ast: GraphAST;
  manifest: PolyglotExecutionManifest;
} {
  const astData = JSON.parse(fragment.astJson) as {
    nodes: Array<[string, ASTNode]>;
    edges: ASTEdge[];
  };

  const nodes = new Map<string, ASTNode>(astData.nodes);
  const ast: GraphAST = { nodes, edges: astData.edges };
  const manifest = JSON.parse(
    fragment.manifestJson
  ) as PolyglotExecutionManifest;

  return { ast, manifest };
}

/**
 * Serialize a GraphAST for storage.
 */
export function serializeAst(ast: GraphAST): string {
  return JSON.stringify({
    nodes: [...ast.nodes.entries()],
    edges: ast.edges,
  });
}

/**
 * Scan a directory and store all topology fragments in D1.
 * This is the build step that pre-parses everything.
 */
export async function scanAndStoreRepo(
  repoRoot: string,
  db: D1Database,
  options?: {
    extensions?: string[];
    exclude?: string[];
    onChange?: (filePath: string, language: string, functions: number) => void;
  }
): Promise<{
  filesScanned: number;
  fragmentsStored: number;
  languagesCovered: string[];
}> {
  const { analyzePolyglotSourceForExecution } = await import(
    './polyglot-bridge.js'
  );

  const extensions = options?.extensions ?? [
    '.py',
    '.go',
    '.rs',
    '.rb',
    '.js',
    '.ts',
    '.jsx',
    '.tsx',
    '.java',
    '.c',
    '.cpp',
    '.lua',
    '.php',
    '.swift',
    '.kt',
    '.scala',
    '.hs',
    '.ml',
    '.zig',
    '.sh',
    '.exs',
    '.cs',
  ];
  const exclude = new Set(
    options?.exclude ?? [
      'node_modules',
      '.git',
      'target',
      'dist',
      'build',
      '__pycache__',
      'vendor',
      '.next',
      'coverage',
      'out',
    ]
  );

  const files = collectSourceFiles(repoRoot, extensions, exclude);
  const languages = new Set<string>();
  let fragmentsStored = 0;

  for (const filePath of files) {
    try {
      const source = fs.readFileSync(filePath, 'utf8');
      const hash = contentHash(source);
      const relativePath = path.relative(repoRoot, filePath);

      const analysis = await analyzePolyglotSourceForExecution(filePath);
      languages.add(analysis.language);

      for (const func of analysis.functions) {
        const key = fragmentKey(relativePath, func.functionName);

        const fragment: TopologyFragment = {
          key,
          filePath: relativePath,
          functionName: func.functionName,
          language: analysis.language,
          astJson: serializeAst(func.ast),
          manifestJson: JSON.stringify(analysis.manifest),
          ggSource: func.ggSource,
          signatureJson: '{}',
          contentHash: hash,
          parsedAt: new Date().toISOString(),
        };

        await storeFragment(db, fragment);
        fragmentsStored++;
      }

      options?.onChange?.(
        relativePath,
        analysis.language,
        analysis.functions.length
      );
    } catch {
      // Skip files that fail to parse.
    }
  }

  return {
    filesScanned: files.length,
    fragmentsStored,
    languagesCovered: [...languages],
  };
}

function collectSourceFiles(
  dir: string,
  extensions: string[],
  exclude: Set<string>
): string[] {
  const files: string[] = [];

  function walk(currentDir: string) {
    let entries: fs.Dirent[];
    try {
      entries = fs.readdirSync(currentDir, { withFileTypes: true });
    } catch {
      return;
    }

    for (const entry of entries) {
      if (exclude.has(entry.name) || entry.name.startsWith('.')) continue;

      const fullPath = path.join(currentDir, entry.name);
      if (entry.isDirectory()) {
        walk(fullPath);
      } else if (entry.isFile()) {
        const ext = path.extname(entry.name);
        if (extensions.includes(ext)) {
          files.push(fullPath);
        }
      }
    }
  }

  walk(dir);
  return files;
}
