// @a0n/.ignored_gnosis/src/polyglot-checkpoint-bridge.ts
/**
 * Polyglot Scanner × Checkpoint Homology Bridge
 *
 * Maps per-function polyglot scanner output (PolyglotFunctionAnalysis)
 * into the universal CheckpointedSystem algebra, so that every project
 * gets a formally-verified health diagnosis computed from actual code
 * topology — not estimates, not heuristics, but real beta1 and deficit
 * from tree-sitter ASTs fed through Betty.
 *
 * The bridge closes the loop:
 *   Lean proves the algebra →
 *   scanner computes the inputs →
 *   bridge maps to algebra →
 *   theorems guarantee the diagnosis is sound →
 *   treatment planner sequences the repair →
 *   repair module injects VENT paths →
 *   re-scan verifies the cure →
 *   loop terminates at TopologicalSaturation (GraphSabbath)
 *
 * Formal backing (6 Lean modules, zero sorry):
 *   CheckpointHomology.lean — self_repair_completeness
 *   PolyglotCheckpointFunctor.lean — scanner → algebra preserves conservation
 *   CancerTreatments.lean — gate-first sequencing, cascade amplification
 *   AgileGnosisTopology.lean — failure as fuel, waterfall collapse proof
 *   EschatonTopology.lean — termination: saturation halts engine
 *   UniversalRepairTopology.lean — 4 mechanisms minimally sufficient,
 *     isomorphic across git/psyche/ecology/macroeconomics
 *
 * @module @a0n/gnosis/polyglot-checkpoint-bridge
 */

import type {
  PolyglotAnalysisResult,
  PolyglotFunctionAnalysis,
  PolyglotDiagnosticFinding,
  PolyglotGhostMassLedger,
} from './polyglot-bridge';
import {
  type CheckpointedSystem,
  type Diagnosis,
  type HealthStatus,
  diagnose,
  godFormulaWeight,
  forecastRepair,
  forecastNeglect,
} from './checkpoint-substrates';
import {
  type GnosisVoidDimension,
  type GnosisVoidDimensionAssessment,
  assessVoidDimension,
} from './void-dimensions';
import {
  type BrunnianResult,
  detectBrunnianCoupling,
  type TopologyModule,
} from './brunnian';

// ═══════════════════════════════════════════════════════════════════════
// Scan Aggregation Types
// ═══════════════════════════════════════════════════════════════════════

/**
 * Four-way health classification for a function.
 *
 *   active  — internal feedback (loops, VENT paths, rejection channels)
 *   covered — no internal feedback but has external coverage (tests)
 *   deaf    — has branching but no feedback of any kind (actually sick)
 *   silent  — pure linear, no branching, no need for feedback
 *
 * Only DEAF functions are actually sick and need repair.
 * Silent functions are healthy by nature. Covered functions are
 * healthy by external verification. Active functions are healthy
 * by internal structure.
 */
export type FunctionHealthClass = 'active' | 'covered' | 'deaf' | 'silent';

/**
 * Per-function metrics extracted from the polyglot scanner + Betty.
 */
export interface FunctionMetrics {
  /** Function name from the scanner. */
  functionName: string;
  /** Source file path. */
  filePath: string;
  /** First Betti number — essential loops, learning capacity. */
  beta1: number;
  /** Topological deficit from Betty's deficit analysis. */
  deficit: number;
  /** Buley complexity measure (topology entropy). */
  buleyMeasure: number;
  /** Void dimensions — FORK branch count. */
  voidDimensions: number;
  /** God Formula weight: w = R - min(v, R) + 1 where R = beta1 + deficit. */
  godWeight: number;
  /** Ghost mass — leaked/unaccounted topology mass. */
  ghostMass: number;
  /** Diagnostic finding count. */
  diagnosticCount: number;
  /** Whether beta1 > 0 (can learn from feedback). */
  isLearning: boolean;
  /** Landauer heat from Betty. */
  landauerHeat: number;
  /** Whether the function has FORK/FOLD branching in its topology. */
  hasBranching: boolean;
  /** Whether a test file exists that covers this function. */
  hasTestCoverage: boolean;
  /**
   * Four-way health classification:
   *   active  — has internal feedback (beta1 > 0)
   *   covered — no internal feedback but has test coverage
   *   deaf    — branching, no feedback, no tests (ACTUALLY SICK)
   *   silent  — no branching at all (healthy by simplicity)
   */
  healthClass: FunctionHealthClass;
}

/**
 * A TwoTermComplex in TypeScript — mirrors CheckpointHomology.TwoTermComplex in Lean.
 * rank1 = active (living), rank0 = destroyed (boundary).
 */
export interface TwoTermComplex {
  rank1: number;
  rank0: number;
}

/**
 * Three independent chain complexes for void dimensions.
 * Each dimension is a separate TwoTermComplex.
 * The composite is the direct sum (rank1_total = sum of rank1s, same for rank0).
 *
 * @see THM-DIRECT-SUM-RANK1-ADDITIVE — rank1 is additive under direct sum
 * @see THM-DIRECT-SUM-RANK0-ADDITIVE — rank0 is additive under direct sum
 */
export interface VoidDimensionProfile {
  /** Inter-module coupling stress. */
  cross_file: TwoTermComplex;
  /** Load-dependent timing stress. */
  temporal: TwoTermComplex;
  /** Query/wire crossing stress. */
  data_plane: TwoTermComplex;
  /** Direct sum of all three dimensions. */
  composite: TwoTermComplex;
  /** Per-dimension assessments from void-dimensions.ts. */
  assessments: GnosisVoidDimensionAssessment[];
}

/**
 * Brunnian coupling interference term for a project.
 */
export interface BrunnianInterference {
  /** Detected Brunnian couplings. */
  couplings: BrunnianResult[];
  /** Total emergent coupling strength across all n-wise knots. */
  totalEmergentStrength: number;
  /** Number of genuine Brunnian links found. */
  brunnianCount: number;
  /** Module IDs involved in at least one Brunnian coupling. */
  affectedModules: string[];
  /**
   * Interference penalty: additional destroyed channels due to
   * Brunnian coupling. When a Brunnian-coupled module is destroyed,
   * its partners are partially degraded.
   */
  interferencePenalty: number;
}

/**
 * Purity profile for a project — classifies functions by topology complexity.
 *
 * A "pure" function has no FORK/FOLD edges (linear Entry→Process→Return).
 * A "branching" function has FORK/FOLD but no cycles (if/else/switch).
 * A "looping" function has PROCESS back-edges (for/while cycles).
 *
 * The purity ratio determines the expected deafness baseline:
 * a library of pure mappers SHOULD have high sigma — pure functions
 * are correctly deaf because they don't iterate or reject.
 *
 * Health thresholds are shifted by purity:
 *   adjustedSigma = rawSigma - purityBaseline
 *   where purityBaseline = pureRatio × 0.8
 *
 * This way, a project with 90% pure functions gets a baseline of 0.72
 * subtracted from its raw sigma before health classification.
 */
export interface PurityProfile {
  /** Functions with no FORK, no loops — simple transformations. */
  pureFunctions: number;
  /** Functions with FORK/FOLD but no loops — branching logic. */
  branchingFunctions: number;
  /** Functions with PROCESS back-edges — iterative logic. */
  loopingFunctions: number;
  /** Total functions classified. */
  totalFunctions: number;
  /** Ratio of pure functions to total (0..1). */
  purityRatio: number;
  /**
   * Expected sigma baseline from purity alone.
   * High-purity projects get a higher baseline (more tolerance).
   * purityBaseline = purityRatio × 0.8
   */
  purityBaseline: number;
  /**
   * Adjusted sigma: rawSigma - purityBaseline, clamped to [0, 1].
   * This is the "real" sickness after accounting for expected purity.
   */
  adjustedSigma: number;
  /**
   * Health classification using adjusted sigma instead of raw sigma.
   */
  adjustedHealth: HealthStatus;
}

/**
 * Aggregated scan result for an entire project.
 */
export interface ProjectScanResult {
  /** Project name / identifier. */
  projectName: string;
  /** Per-function metrics. */
  functions: FunctionMetrics[];
  /** Total function count. */
  functionCount: number;
  /** Sum of beta1 across all functions. */
  totalBeta1: number;
  /** Sum of deficit across all functions. */
  totalDeficit: number;
  /** Sum of God Formula weights. */
  totalGodWeight: number;
  /** Maximum God Formula weight (worst single function). */
  maxGodWeight: number;
  /** Functions with godWeight > hotspot threshold. */
  hotspotCount: number;
  /** Sum of ghost mass across all functions. */
  totalGhostMass: number;
  /** Sum of void dimensions across all functions. */
  totalVoidDimensions: number;
  /** Sum of Buley measures across all functions. */
  totalBuleyMeasure: number;
  /** Total diagnostic findings. */
  totalDiagnostics: number;
  /** Files analyzed. */
  filesAnalyzed: number;
  /** Errors encountered during analysis. */
  errors: string[];
}

/**
 * Full checkpoint profile for a project — the bridge output.
 */
export interface ProjectCheckpointProfile {
  /** Project name. */
  projectName: string;
  /** Raw scan aggregation. */
  scan: ProjectScanResult;
  /** Functions with beta1 > 0 (active rejection channels). */
  activeChannels: number;
  /** Functions with beta1 = 0 (destroyed rejection channels). */
  destroyedChannels: number;
  /** The CheckpointedSystem mapped from the scan. */
  system: CheckpointedSystem;
  /** Diagnosis from checkpoint-substrates.ts. */
  diagnosis: Diagnosis;
  /**
   * Real sigma: normalized stress from actual code topology.
   * sicknessFraction from the CheckpointedSystem.
   */
  sigma: number;
  /**
   * Top functions ordered by God Formula weight (highest = fix first).
   * These have the most deficit and will give the most beta1 gain per repair.
   */
  repairOrdering: RepairCandidate[];
  /**
   * Forecast: what happens if we restore the top candidate.
   */
  forecastIfRepaired: CheckpointedSystem;
  /**
   * Forecast: what happens if we lose one more channel.
   */
  forecastIfNeglected: CheckpointedSystem;
  /**
   * Void dimension profile: 3 independent chain complexes.
   * Shows WHERE the disease lives, not just THAT it exists.
   */
  voidProfile: VoidDimensionProfile;
  /**
   * Brunnian coupling interference.
   * n-wise emergent knots that pairwise analysis misses.
   */
  brunnianInterference: BrunnianInterference;
  /**
   * Purity profile — classifies functions as pure/branching/looping.
   * Used to calibrate health thresholds for the project type.
   */
  purityProfile: PurityProfile;
}

/**
 * A function ranked for repair priority by the God Formula.
 */
export interface RepairCandidate {
  functionName: string;
  filePath: string;
  godWeight: number;
  beta1: number;
  deficit: number;
  ghostMass: number;
}

// ═══════════════════════════════════════════════════════════════════════
// Configuration
// ═══════════════════════════════════════════════════════════════════════

/** God weight threshold for hotspot classification. */
const HOTSPOT_THRESHOLD = 10;

/** Maximum repair candidates to include in ordering. */
const MAX_REPAIR_CANDIDATES = 10;

// ═══════════════════════════════════════════════════════════════════════
// Core Bridge Functions
// ═══════════════════════════════════════════════════════════════════════

/**
 * Extract per-function metrics from a PolyglotFunctionAnalysis.
 *
 * Diversity theory: throw the kitchen sink at beta1 estimation.
 * Betty's b1 from FORK/FOLD topology is the primary signal, but
 * supplementary signals boost it when the Rust scanner emits loops
 * as PROCESS back-edges rather than proper loop edges:
 *
 *   1. Betty b1 (FORK/FOLD cycles in GG topology)
 *   2. VOID_METACOG_MISSING_CONVERGENCE diagnostics (PROCESS back-edges
 *      = for/while loops the scanner detected but didn't emit as loops)
 *   3. ETHICS_MISSING_VENT_PATH absent (if a VENT exists, the function
 *      has a rejection channel — it can learn from failure)
 *
 * The effective beta1 is: betty_b1 + cycle_count
 * Each PROCESS back-edge cycle contributes 1 to learning capacity
 * because a loop CAN learn (it iterates until a condition is met).
 */
/**
 * Test coverage detection — check if a function has external verification.
 *
 * Three signals, checked in order:
 *   1. LCOV data (if provided) — line-level coverage, most precise
 *   2. Test file exists — foo.test.ts exists for foo.ts
 *   3. Test import — the function name appears in a .test. file
 *
 * TODO: Build topology-aware coverage that checks whether a test
 * exercises the function's specific FORK branches, not just its lines.
 * That would be a gnosis-native coverage signal — better than LCOV
 * because it knows which branches are deaf vs. tested.
 */
function detectTestCoverage(
  functionName: string,
  filePath: string,
  testFileIndex: Set<string>,
): boolean {
  // Check if a corresponding test file exists
  // foo.ts → foo.test.ts, foo.spec.ts
  const base = filePath.replace(/\.(ts|tsx|js|jsx)$/, '');
  if (testFileIndex.has(base + '.test.ts') ||
      testFileIndex.has(base + '.test.tsx') ||
      testFileIndex.has(base + '.spec.ts') ||
      testFileIndex.has(base + '.spec.tsx') ||
      testFileIndex.has(base + '.test.js') ||
      testFileIndex.has(base + '.spec.js')) {
    return true;
  }

  // Check if the file IS a test file (functions inside test files are covered)
  if (filePath.includes('.test.') || filePath.includes('.spec.') ||
      filePath.includes('__tests__')) {
    return true;
  }

  return false;
}

/**
 * Build a set of test file paths for quick lookup.
 */
export function buildTestFileIndex(
  fileResults: Array<{ filePath: string }>,
): Set<string> {
  const index = new Set<string>();
  for (const { filePath } of fileResults) {
    if (filePath.includes('.test.') || filePath.includes('.spec.') ||
        filePath.includes('__tests__')) {
      index.add(filePath);
    }
  }
  return index;
}

/**
 * Classify a function into one of four health classes.
 *
 *   active  — has internal feedback (beta1 > 0 from loops/VENT)
 *   covered — no internal feedback but has test coverage
 *   deaf    — has branching but no feedback of any kind (ACTUALLY SICK)
 *   silent  — no branching at all (healthy by simplicity)
 *
 * Only DEAF functions are repair targets. The rest are healthy.
 */
function classifyFunctionHealth(
  beta1: number,
  hasBranching: boolean,
  hasTestCoverage: boolean,
): FunctionHealthClass {
  if (beta1 > 0) return 'active';
  if (!hasBranching) return 'silent';
  if (hasTestCoverage) return 'covered';
  return 'deaf';
}

/**
 * Handles the gnosis extract Function Metrics workflow.
 */
export function extractFunctionMetrics(
  func: PolyglotFunctionAnalysis,
  filePath: string,
  testFileIndex?: Set<string>,
): FunctionMetrics {
  const bettyBeta1 = func.bettyResult.b1;

  // Count PROCESS back-edge cycles from Betty diagnostics.
  const cycleCount = func.bettyResult.diagnostics.filter(
    (d) => d.code === 'VOID_METACOG_MISSING_CONVERGENCE',
  ).length;

  // Check for VENT edges in the topology — rejection channels.
  const hasVent = func.bettyResult.ast
    ? func.bettyResult.ast.edges.some((e) => e.type === 'VENT')
    : false;
  const ventBonus = hasVent ? 1 : 0;

  // Effective beta1: Betty's b1 + loop cycles + vent bonus.
  const beta1 = bettyBeta1 + cycleCount + ventBonus;

  // Check for branching (FORK edges in topology)
  const hasBranching = func.bettyResult.ast
    ? func.bettyResult.ast.edges.some((e) => e.type === 'FORK')
    : false;

  // Check for test coverage
  const hasTestCoverage = testFileIndex
    ? detectTestCoverage(func.functionName, filePath, testFileIndex)
    : false;

  // Four-way health classification
  const healthClass = classifyFunctionHealth(beta1, hasBranching, hasTestCoverage);

  const deficit = func.bettyResult.deficit?.totalDeficit ?? 0;
  const totalRounds = beta1 + deficit;
  const godWeight = godFormulaWeight(totalRounds, deficit);

  return {
    functionName: func.functionName,
    filePath,
    beta1,
    deficit,
    buleyMeasure: func.bettyResult.buleyMeasure,
    voidDimensions: func.bettyResult.voidDimensions,
    godWeight,
    ghostMass: func.ghostMassLedger.ghost_mass,
    diagnosticCount: func.diagnostics.length,
    isLearning: beta1 > 0,
    landauerHeat: func.bettyResult.landauerHeat,
    hasBranching,
    hasTestCoverage,
    healthClass,
  };
}

/**
 * Aggregate multiple file analysis results into a ProjectScanResult.
 */
export function aggregateFindings(
  projectName: string,
  fileResults: Array<{ filePath: string; result: PolyglotAnalysisResult }>,
): ProjectScanResult {
  const functions: FunctionMetrics[] = [];
  const errors: string[] = [];

  // Build test file index for coverage detection
  const testFileIndex = buildTestFileIndex(fileResults);

  for (const { filePath, result } of fileResults) {
    errors.push(...result.errors);
    for (const func of result.functions) {
      functions.push(extractFunctionMetrics(func, filePath, testFileIndex));
    }
  }

  const totalBeta1 = functions.reduce((s, f) => s + f.beta1, 0);
  const totalDeficit = functions.reduce((s, f) => s + f.deficit, 0);
  const totalGodWeight = functions.reduce((s, f) => s + f.godWeight, 0);
  const maxGodWeight = functions.reduce((m, f) => Math.max(m, f.godWeight), 0);
  const hotspotCount = functions.filter(
    (f) => f.godWeight > HOTSPOT_THRESHOLD,
  ).length;
  const totalGhostMass = functions.reduce((s, f) => s + f.ghostMass, 0);
  const totalVoidDimensions = functions.reduce(
    (s, f) => s + f.voidDimensions,
    0,
  );
  const totalBuleyMeasure = functions.reduce((s, f) => s + f.buleyMeasure, 0);
  const totalDiagnostics = functions.reduce(
    (s, f) => s + f.diagnosticCount,
    0,
  );

  return {
    projectName,
    functions,
    functionCount: functions.length,
    totalBeta1,
    totalDeficit,
    totalGodWeight,
    maxGodWeight,
    hotspotCount,
    totalGhostMass,
    totalVoidDimensions,
    totalBuleyMeasure,
    totalDiagnostics,
    filesAnalyzed: fileResults.length,
    errors,
  };
}

/**
 * Map a ProjectScanResult to a CheckpointedSystem.
 *
 * The mapping:
 *   - active channels = functions with beta1 > 0 (can learn from feedback)
 *   - destroyed channels = functions with beta1 = 0 (deaf — no loops)
 *
 * This preserves the homological invariant:
 *   active + destroyed = functionCount (scan_preserves_conservation)
 *
 * The resulting system's sicknessFraction is the real sigma.
 */
export function scanToCheckpointedSystem(
  scan: ProjectScanResult,
): CheckpointedSystem {
  const active = scan.functions.filter((f) => f.isLearning).length;
  const destroyed = scan.functions.filter((f) => !f.isLearning).length;

  // Use the imported buildSystem logic inline (it's not exported)
  const beta1 = active;
  const deficit = destroyed;
  const healthyBeta1 = beta1 + deficit;
  const total = beta1 + deficit;

  return {
    active,
    destroyed,
    beta1,
    deficit,
    healthyBeta1,
    isDeaf: active === 0,
    isLearning: active > 0,
    sicknessFraction: total > 0 ? deficit / total : 0,
  };
}

/**
 * Compute repair ordering: functions sorted by God Formula weight descending.
 * Highest weight = most deficit relative to capacity = fix first.
 */
export function computeRepairOrdering(
  scan: ProjectScanResult,
  limit: number = MAX_REPAIR_CANDIDATES,
): RepairCandidate[] {
  return [...scan.functions]
    .sort((a, b) => b.godWeight - a.godWeight)
    .slice(0, limit)
    .map((f) => ({
      functionName: f.functionName,
      filePath: f.filePath,
      godWeight: f.godWeight,
      beta1: f.beta1,
      deficit: f.deficit,
      ghostMass: f.ghostMass,
    }));
}

// ═══════════════════════════════════════════════════════════════════════
// Step 5: Void Dimension Profiles
// ═══════════════════════════════════════════════════════════════════════

/**
 * Build a TwoTermComplex from visible/invisible crossing counts.
 * Visible crossings = active (we can see and fix them).
 * Invisible crossings = destroyed (hidden, causing silent failures).
 */
function crossingsToComplex(
  visible: number,
  invisible: number,
): TwoTermComplex {
  return { rank1: visible, rank0: invisible };
}

/**
 * Direct sum of TwoTermComplexes.
 * @see THM-DIRECT-SUM-RANK1-ADDITIVE, THM-DIRECT-SUM-RANK0-ADDITIVE
 */
export function directSum(a: TwoTermComplex, b: TwoTermComplex): TwoTermComplex {
  return { rank1: a.rank1 + b.rank1, rank0: a.rank0 + b.rank0 };
}

/**
 * Compute the void dimension profile for a project.
 *
 * Each of the three void dimensions (cross_file, temporal, data_plane)
 * becomes an independent TwoTermComplex. The composite is their direct sum.
 *
 * The assessVoidDimension function from void-dimensions.ts gives us
 * visible/invisible crossing counts per dimension. We map:
 *   - visibleCrossings → rank1 (active — we can see and fix them)
 *   - invisibleCrossings → rank0 (destroyed — hidden damage)
 *
 * The CheckpointHomology.directSum theorem proves that composite health
 * is the sum of the three dimensions. A project can be healthy in
 * cross_file but stressed in temporal.
 */
export function computeVoidProfile(
  scan: ProjectScanResult,
  intermoduleEdges: number = 0,
  queueDepth: number = 0,
  dataCrossings: number = 0,
): VoidDimensionProfile {
  // Visible crossings from the scanner's voidDimensions (FORK branches)
  const visibleCrossings = scan.totalVoidDimensions;

  // Distribute visible crossings heuristically across dimensions
  // (until the scanner produces per-dimension counts)
  const crossFileVisible = Math.ceil(visibleCrossings * 0.5);
  const temporalVisible = Math.ceil(visibleCrossings * 0.3);
  const dataPlaneVisible = visibleCrossings - crossFileVisible - temporalVisible;

  const crossFileAssessment = assessVoidDimension({
    dimension: 'cross_file',
    visibleCrossings: crossFileVisible,
    intermoduleEdges,
  });

  const temporalAssessment = assessVoidDimension({
    dimension: 'temporal',
    visibleCrossings: temporalVisible,
    queueDepth,
  });

  const dataPlaneAssessment = assessVoidDimension({
    dimension: 'data_plane',
    visibleCrossings: dataPlaneVisible,
    dataCrossings,
  });

  const cross_file = crossingsToComplex(
    crossFileAssessment.visibleCrossings,
    crossFileAssessment.invisibleCrossings,
  );
  const temporal = crossingsToComplex(
    temporalAssessment.visibleCrossings,
    temporalAssessment.invisibleCrossings,
  );
  const data_plane = crossingsToComplex(
    dataPlaneAssessment.visibleCrossings,
    dataPlaneAssessment.invisibleCrossings,
  );

  const composite = directSum(directSum(cross_file, temporal), data_plane);

  return {
    cross_file,
    temporal,
    data_plane,
    composite,
    assessments: [crossFileAssessment, temporalAssessment, dataPlaneAssessment],
  };
}

// ═══════════════════════════════════════════════════════════════════════
// Step 6: Brunnian Coupling as Interference
// ═══════════════════════════════════════════════════════════════════════

/**
 * Detect Brunnian coupling interference across a project's functions.
 *
 * Groups functions by file, treats each file as a TopologyModule,
 * and runs detectBrunnianCoupling on all 3-wise combinations.
 *
 * The interference penalty is the total emergent strength — additional
 * destroyed channels that pairwise analysis would miss.
 */
export function computeBrunnianInterference(
  scan: ProjectScanResult,
  modules?: TopologyModule[],
): BrunnianInterference {
  // If modules aren't provided, we can't run Brunnian detection
  // (it requires actual .gg topology sources, not just metrics).
  // Return empty interference.
  if (!modules || modules.length < 3) {
    return {
      couplings: [],
      totalEmergentStrength: 0,
      brunnianCount: 0,
      affectedModules: [],
      interferencePenalty: 0,
    };
  }

  // Run Brunnian detection on all 3-wise combinations
  const couplings: BrunnianResult[] = [];
  for (let i = 0; i < modules.length; i++) {
    for (let j = i + 1; j < modules.length; j++) {
      for (let k = j + 1; k < modules.length; k++) {
        const result = detectBrunnianCoupling([
          modules[i],
          modules[j],
          modules[k],
        ]);
        if (result.isBrunnian) {
          couplings.push(result);
        }
      }
    }
  }

  const totalEmergentStrength = couplings.reduce(
    (s, c) => s + c.emergentStrength,
    0,
  );

  const affectedModuleSet = new Set<string>();
  for (const c of couplings) {
    for (const m of c.modules) {
      affectedModuleSet.add(m);
    }
  }

  // Interference penalty: each Brunnian knot adds partial destruction
  // to its constituent modules. The penalty is the number of affected
  // modules that would experience cascading degradation.
  const interferencePenalty = Math.min(
    scan.functionCount,
    Math.ceil(totalEmergentStrength * 0.5),
  );

  return {
    couplings,
    totalEmergentStrength,
    brunnianCount: couplings.length,
    affectedModules: Array.from(affectedModuleSet),
    interferencePenalty,
  };
}

// ═══════════════════════════════════════════════════════════════════════
// Purity Profile — calibrate health thresholds by project type
// ═══════════════════════════════════════════════════════════════════════

/**
 * Compute the purity profile for a project from its file results.
 *
 * Classifies each function as:
 *   - pure: no FORK edges, no cycle diagnostics (linear topology)
 *   - branching: has FORK/FOLD but no cycles (if/else/switch)
 *   - looping: has PROCESS back-edge cycles (for/while)
 *
 * The purity ratio determines the expected deafness baseline:
 *   purityBaseline = purityRatio × 0.8
 *   adjustedSigma = max(0, rawSigma - purityBaseline)
 *
 * A library of pure mappers (purity 0.9) gets baseline 0.72,
 * so raw sigma 0.9 becomes adjusted sigma 0.18 → healthy.
 */
export function computePurityProfile(
  scan: ProjectScanResult,
  fileResults: Array<{ filePath: string; result: PolyglotAnalysisResult }>,
  rawSigma: number,
): PurityProfile {
  let pureFunctions = 0;
  let branchingFunctions = 0;
  let loopingFunctions = 0;

  for (const { result } of fileResults) {
    for (const func of result.functions) {
      const edges = func.bettyResult.ast?.edges ?? [];
      const hasFork = edges.some((e) => e.type === 'FORK');
      const hasCycle = func.bettyResult.diagnostics.some(
        (d) => d.code === 'VOID_METACOG_MISSING_CONVERGENCE',
      );

      if (hasCycle) {
        loopingFunctions++;
      } else if (hasFork) {
        branchingFunctions++;
      } else {
        pureFunctions++;
      }
    }
  }

  const totalFunctions = pureFunctions + branchingFunctions + loopingFunctions;
  const purityRatio = totalFunctions > 0 ? pureFunctions / totalFunctions : 0;

  // Purity baseline: how much sigma to "forgive" based on expected purity.
  // Pure functions are CORRECTLY deaf — they don't need feedback loops.
  // A project with 90% pure functions should tolerate 72% deafness.
  const purityBaseline = purityRatio * 0.8;

  // Adjusted sigma: the real sickness after subtracting expected purity.
  const adjustedSigma = Math.max(0, Math.min(1, rawSigma - purityBaseline));

  // Classify health using adjusted sigma.
  let adjustedHealth: HealthStatus;
  if (adjustedSigma >= 1.0) {
    adjustedHealth = 'necrotic';
  } else if (adjustedSigma >= 0.6) {
    adjustedHealth = 'cancerous';
  } else if (adjustedSigma >= 0.3) {
    adjustedHealth = 'inflamed';
  } else if (adjustedSigma >= 0.1) {
    adjustedHealth = 'stressed';
  } else {
    adjustedHealth = 'healthy';
  }

  return {
    pureFunctions,
    branchingFunctions,
    loopingFunctions,
    totalFunctions,
    purityRatio,
    purityBaseline,
    adjustedSigma,
    adjustedHealth,
  };
}

// ═══════════════════════════════════════════════════════════════════════
// Primary Entry Point (updated with void profile + Brunnian + purity)
// ═══════════════════════════════════════════════════════════════════════

/**
 * Full bridge: aggregate scanner output → checkpoint profile with diagnosis.
 *
 * This is the primary entry point. Given a project name and a set of
 * file-level polyglot analysis results, it produces a complete
 * ProjectCheckpointProfile with:
 *   - Real sigma from actual code topology
 *   - Health classification from CheckpointHomology
 *   - Repair ordering from the God Formula
 *   - Forecasts for repair and neglect trajectories
 *   - Void dimension profile (where the disease lives)
 *   - Brunnian interference (n-wise emergent coupling)
 */
export function diagnoseProject(
  projectName: string,
  fileResults: Array<{ filePath: string; result: PolyglotAnalysisResult }>,
  options?: {
    brunnianModules?: TopologyModule[];
    intermoduleEdges?: number;
    queueDepth?: number;
    dataCrossings?: number;
  },
): ProjectCheckpointProfile {
  const scan = aggregateFindings(projectName, fileResults);
  const system = scanToCheckpointedSystem(scan);
  const diagnosis = diagnose(system);
  const repairOrdering = computeRepairOrdering(scan);
  const voidProfile = computeVoidProfile(
    scan,
    options?.intermoduleEdges,
    options?.queueDepth,
    options?.dataCrossings,
  );
  const brunnianInterference = computeBrunnianInterference(
    scan,
    options?.brunnianModules,
  );
  const purityProfile = computePurityProfile(
    scan,
    fileResults,
    system.sicknessFraction,
  );

  return {
    projectName,
    scan,
    activeChannels: system.active,
    destroyedChannels: system.destroyed,
    system,
    diagnosis,
    sigma: system.sicknessFraction,
    repairOrdering,
    forecastIfRepaired: forecastRepair(system, 1),
    forecastIfNeglected: forecastNeglect(system, 1),
    voidProfile,
    brunnianInterference,
    purityProfile,
  };
}

// ═══════════════════════════════════════════════════════════════════════
// Utility: Quick diagnosis from raw counts (no scanner needed)
// ═══════════════════════════════════════════════════════════════════════

/**
 * Quick bridge for when you already have active/destroyed counts.
 * Useful for testing or when scan results are cached.
 */
export function diagnoseFromCounts(
  projectName: string,
  activeFunctions: number,
  destroyedFunctions: number,
): Pick<
  ProjectCheckpointProfile,
  | 'projectName'
  | 'activeChannels'
  | 'destroyedChannels'
  | 'system'
  | 'diagnosis'
  | 'sigma'
  | 'forecastIfRepaired'
  | 'forecastIfNeglected'
> {
  const beta1 = activeFunctions;
  const deficit = destroyedFunctions;
  const healthyBeta1 = beta1 + deficit;
  const total = beta1 + deficit;

  const system: CheckpointedSystem = {
    active: activeFunctions,
    destroyed: destroyedFunctions,
    beta1,
    deficit,
    healthyBeta1,
    isDeaf: activeFunctions === 0,
    isLearning: activeFunctions > 0,
    sicknessFraction: total > 0 ? deficit / total : 0,
  };

  return {
    projectName,
    activeChannels: activeFunctions,
    destroyedChannels: destroyedFunctions,
    system,
    diagnosis: diagnose(system),
    sigma: system.sicknessFraction,
    forecastIfRepaired: forecastRepair(system, 1),
    forecastIfNeglected: forecastNeglect(system, 1),
  };
}
