// @a0n/.ignored_gnosis/src/polyglot-watch.ts
/**
 * gnode watch -- polyglot hot-reload during development.
 *
 * Watches source files, re-analyzes on change via the persistent scanner
 * service, Betty-verifies the new topology, and hot-swaps changed nodes.
 */

import * as fs from 'node:fs';
import * as path from 'node:path';
import { analyzePolyglotSourceForExecution } from './polyglot-bridge';
import { executePolyglotWithGnosis } from './polyglot-execution-bridge';
import type { GraphAST } from './betty/compiler';
import type { PolyglotExecutionManifest } from './polyglot-execution-bridge';

/**
 * Watch configuration.
 */
export interface WatchConfig {
  /** Files to watch. */
  files: string[];
  /** Function to execute on change (optional -- if set, re-runs after each change). */
  exportName?: string;
  /** Input for execution. */
  input?: unknown;
  /** Debounce interval in ms (default: 300). */
  debounceMs?: number;
  /** Callback when a file changes and is re-analyzed. */
  onChange?: (event: WatchChangeEvent) => void;
  /** Callback on execution result. */
  onResult?: (result: WatchExecutionResult) => void;
  /** Callback on error. */
  onError?: (error: Error) => void;
}

export interface WatchChangeEvent {
  filePath: string;
  language: string;
  functions: string[];
  analysisMs: number;
  topologyChanged: boolean;
}

export interface WatchExecutionResult {
  payload: unknown;
  logs: string;
  executionMs: number;
}

interface WatchedFileState {
  filePath: string;
  lastModified: number;
  lastFingerprint: string;
  language: string;
  functions: string[];
  ast?: GraphAST;
  manifest?: PolyglotExecutionManifest;
}

/**
 * Start watching files for changes.
 * Returns a stop function to terminate the watcher.
 */
export function startWatch(config: WatchConfig): { stop: () => void } {
  const debounceMs = config.debounceMs ?? 300;
  const fileStates = new Map<string, WatchedFileState>();
  const watchers: fs.FSWatcher[] = [];
  let debounceTimer: ReturnType<typeof setTimeout> | null = null;
  let running = false;
  let stopped = false;

  // Initial analysis of all files.
  const initPromise = (async () => {
    for (const filePath of config.files) {
      await analyzeAndUpdateState(filePath);
    }
  })();

  // Set up file watchers.
  for (const filePath of config.files) {
    try {
      const watcher = fs.watch(filePath, (eventType) => {
        if (stopped || eventType !== 'change') return;
        scheduleReanalysis(filePath);
      });
      watchers.push(watcher);
    } catch {
      // File might not exist yet -- watch the directory.
      const dir = path.dirname(filePath);
      const basename = path.basename(filePath);
      try {
        const watcher = fs.watch(dir, (eventType, filename) => {
          if (stopped || filename !== basename) return;
          scheduleReanalysis(filePath);
        });
        watchers.push(watcher);
      } catch {
        config.onError?.(new Error(`Cannot watch ${filePath}`));
      }
    }
  }

  function scheduleReanalysis(filePath: string) {
    if (debounceTimer) clearTimeout(debounceTimer);
    debounceTimer = setTimeout(() => handleChange(filePath), debounceMs);
  }

  async function handleChange(filePath: string) {
    if (running || stopped) return;
    running = true;

    try {
      const event = await analyzeAndUpdateState(filePath);
      if (event) {
        config.onChange?.(event);

        // Re-execute if configured.
        if (config.exportName || config.input !== undefined) {
          const state = fileStates.get(filePath);
          if (state?.ast && state.manifest) {
            const execStarted = performance.now();
            try {
              const result = await executePolyglotWithGnosis({
                ast: state.ast,
                manifest: state.manifest,
                input: config.input,
              });
              config.onResult?.({
                payload: result.payload,
                logs: result.logs,
                executionMs: performance.now() - execStarted,
              });
            } catch (err: unknown) {
              config.onError?.(
                err instanceof Error ? err : new Error(String(err))
              );
            }
          }
        }
      }
    } finally {
      running = false;
    }
  }

  async function analyzeAndUpdateState(
    filePath: string
  ): Promise<WatchChangeEvent | null> {
    const absolutePath = path.resolve(filePath);
    let stat: fs.Stats;
    try {
      stat = fs.statSync(absolutePath);
    } catch {
      return null; // File deleted or not yet created.
    }

    const existing = fileStates.get(absolutePath);
    if (existing && existing.lastModified >= stat.mtimeMs) {
      return null; // No change.
    }

    const started = performance.now();
    const analysis = await analyzePolyglotSourceForExecution(absolutePath);
    const analysisMs = performance.now() - started;

    // Compute fingerprint from function names + gg sources.
    const fingerprint = analysis.functions
      .map((f) => `${f.functionName}:${f.ggSource.length}`)
      .join('|');

    const topologyChanged = existing?.lastFingerprint !== fingerprint;

    const state: WatchedFileState = {
      filePath: absolutePath,
      lastModified: stat.mtimeMs,
      lastFingerprint: fingerprint,
      language: analysis.language,
      functions: analysis.functions.map((f) => f.functionName),
      ast: analysis.functions[0]?.ast,
      manifest: analysis.manifest,
    };
    fileStates.set(absolutePath, state);

    return {
      filePath: absolutePath,
      language: analysis.language,
      functions: state.functions,
      analysisMs,
      topologyChanged,
    };
  }

  return {
    stop() {
      stopped = true;
      if (debounceTimer) clearTimeout(debounceTimer);
      for (const watcher of watchers) {
        watcher.close();
      }
      watchers.length = 0;
    },
  };
}
