// @a0n/.ignored_gnosis/src/polyglot-memory.ts
/**
 * Polyglot language strategy memory -- a CRDT journal that tracks which
 * language implementations worked best for each function across invocations.
 *
 * Similar to x-gnosis's CommunityCacheJournal: an append-only history with
 * exponential decay, so recent observations matter more than old ones.
 *
 * The memory answers questions like:
 * - "Last time we ran parse_csv, Python took 120ms and Go took 15ms"
 * - "The Rust implementation of transform_data had a 30% error rate"
 * - "For compute-heavy functions, Go is 4x faster than Python in this codebase"
 *
 * Stored as a QDoc CRDT for distributed sharing via AeonRelay.
 */

import { QDoc } from './crdt/qdoc';

/**
 * A single observation: one function executed in one language.
 */
export interface LanguageObservation {
  /** Function name. */
  functionName: string;
  /** Language used. */
  language: string;
  /** Source file path. */
  filePath: string;
  /** Execution duration in ms. */
  durationMs: number;
  /** Whether execution succeeded. */
  success: boolean;
  /** Error message if failed. */
  error?: string;
  /** Timestamp (ISO 8601). */
  timestamp: string;
  /** Beta cost of the topology. */
  betaCost?: number;
  /** Cache status (hit/miss). */
  cacheStatus?: 'hit' | 'miss';
}

/**
 * Aggregated strategy score for a language+function pair.
 */
export interface StrategyScore {
  /** Language. */
  language: string;
  /** Function name. */
  functionName: string;
  /** Number of observations. */
  observations: number;
  /** Exponentially weighted mean duration (ms). Recent observations weighted more. */
  weightedDurationMs: number;
  /** Success rate (0.0 - 1.0). */
  successRate: number;
  /** Overall fitness score (higher = better). */
  fitnessScore: number;
  /** Last observation timestamp. */
  lastSeen: string;
}

/**
 * Decay configuration.
 */
export interface DecayConfig {
  /** Half-life in observations (default: 10 -- after 10 observations, old data has half weight). */
  halfLife: number;
  /** Maximum observations to retain per function+language pair. */
  maxObservations: number;
  /** Prune observations older than this many days. */
  maxAgeDays: number;
}

const DEFAULT_DECAY: DecayConfig = {
  halfLife: 10,
  maxObservations: 50,
  maxAgeDays: 7,
};

/**
 * Polyglot strategy memory -- append-only CRDT journal with exponential decay.
 */
export class PolyglotStrategyMemory {
  private readonly qdoc: QDoc;
  private readonly decay: DecayConfig;
  /** In-memory observation buffer, keyed by `function:language`. */
  private observations = new Map<string, LanguageObservation[]>();

  constructor(options?: { guid?: string; decay?: Partial<DecayConfig> }) {
    this.qdoc = new QDoc({
      guid: options?.guid ?? 'gnode-polyglot-strategy-memory',
    });
    this.decay = { ...DEFAULT_DECAY, ...options?.decay };
  }

  /**
   * Record an observation.
   */
  record(observation: LanguageObservation): void {
    const key = `${observation.functionName}:${observation.language}`;
    if (!this.observations.has(key)) {
      this.observations.set(key, []);
    }
    const buffer = this.observations.get(key)!;
    buffer.push(observation);

    // Prune if over max.
    if (buffer.length > this.decay.maxObservations) {
      buffer.shift();
    }

    // Persist to QDoc.
    this.persistToQDoc(key, buffer);
  }

  /**
   * Get the aggregated strategy score for a function+language pair.
   */
  score(functionName: string, language: string): StrategyScore | null {
    const key = `${functionName}:${language}`;
    const buffer = this.observations.get(key);
    if (!buffer || buffer.length === 0) return null;

    return this.computeScore(functionName, language, buffer);
  }

  /**
   * Rank all languages for a function by fitness score.
   */
  rank(functionName: string): StrategyScore[] {
    const scores: StrategyScore[] = [];

    for (const [key, buffer] of this.observations) {
      const [fn, lang] = key.split(':');
      if (fn !== functionName || !lang) continue;
      const s = this.computeScore(functionName, lang, buffer);
      if (s) scores.push(s);
    }

    return scores.sort((a, b) => b.fitnessScore - a.fitnessScore);
  }

  /**
   * Get the recommended language for a function based on historical performance.
   */
  recommend(functionName: string): string | null {
    const ranked = this.rank(functionName);
    return ranked.length > 0 ? ranked[0]!.language : null;
  }

  /**
   * Get all observations for a function (across all languages).
   */
  history(functionName: string): LanguageObservation[] {
    const all: LanguageObservation[] = [];
    for (const [key, buffer] of this.observations) {
      if (key.startsWith(`${functionName}:`)) {
        all.push(...buffer);
      }
    }
    return all.sort(
      (a, b) =>
        new Date(b.timestamp).getTime() - new Date(a.timestamp).getTime()
    );
  }

  /**
   * Get a summary of all tracked functions and their best languages.
   */
  summary(): Array<{
    functionName: string;
    bestLanguage: string;
    alternatives: Array<{ language: string; score: number }>;
    totalObservations: number;
  }> {
    const functionNames = new Set<string>();
    for (const key of this.observations.keys()) {
      const fn = key.split(':')[0];
      if (fn) functionNames.add(fn);
    }

    return [...functionNames].map((fn) => {
      const ranked = this.rank(fn);
      const totalObservations = ranked.reduce(
        (sum, s) => sum + s.observations,
        0
      );
      return {
        functionName: fn,
        bestLanguage: ranked[0]?.language ?? 'unknown',
        alternatives: ranked.map((s) => ({
          language: s.language,
          score: s.fitnessScore,
        })),
        totalObservations,
      };
    });
  }

  /**
   * Prune old observations based on decay config.
   */
  prune(): void {
    const cutoff = new Date();
    cutoff.setDate(cutoff.getDate() - this.decay.maxAgeDays);
    const cutoffTs = cutoff.getTime();

    for (const [key, buffer] of this.observations) {
      const pruned = buffer.filter(
        (obs) => new Date(obs.timestamp).getTime() >= cutoffTs
      );
      if (pruned.length === 0) {
        this.observations.delete(key);
      } else {
        this.observations.set(key, pruned);
      }
    }
  }

  /**
   * Serialize the full memory state for persistence.
   */
  serialize(): string {
    const data: Record<string, LanguageObservation[]> = {};
    for (const [key, buffer] of this.observations) {
      data[key] = buffer;
    }
    return JSON.stringify(data);
  }

  /**
   * Restore memory state from serialized data.
   */
  restore(serialized: string): void {
    const data = JSON.parse(serialized) as Record<
      string,
      LanguageObservation[]
    >;
    this.observations.clear();
    for (const [key, buffer] of Object.entries(data)) {
      this.observations.set(key, buffer);
    }
  }

  /**
   * Get the underlying QDoc for relay synchronization.
   */
  getQDoc(): QDoc {
    return this.qdoc;
  }

  // ─── Private ─────────────────────────────────────────────────────────────

  private computeScore(
    functionName: string,
    language: string,
    buffer: LanguageObservation[]
  ): StrategyScore {
    const lambda = Math.LN2 / this.decay.halfLife;
    let weightedDuration = 0;
    let totalWeight = 0;
    let successes = 0;

    for (let i = 0; i < buffer.length; i++) {
      const obs = buffer[i]!;
      // Exponential decay: more recent observations have more weight.
      const age = buffer.length - 1 - i; // 0 = most recent
      const weight = Math.exp(-lambda * age);

      weightedDuration += obs.durationMs * weight;
      totalWeight += weight;

      if (obs.success) successes += 1;
    }

    const weightedDurationMs =
      totalWeight > 0 ? weightedDuration / totalWeight : 0;
    const successRate = buffer.length > 0 ? successes / buffer.length : 0;

    // Fitness = success_rate / (1 + log(duration)).
    // High success + low duration = high fitness.
    const fitnessScore =
      successRate / (1 + Math.log1p(weightedDurationMs / 100));

    return {
      language,
      functionName,
      observations: buffer.length,
      weightedDurationMs,
      successRate,
      fitnessScore,
      lastSeen: buffer[buffer.length - 1]?.timestamp ?? '',
    };
  }

  private persistToQDoc(key: string, buffer: LanguageObservation[]): void {
    // Store the latest observation in the QDoc map.
    // The full buffer is in-memory; the QDoc persists the latest state
    // for relay synchronization across machines.
    const latest = buffer[buffer.length - 1];
    if (!latest) return;

    const score = this.computeScore(
      latest.functionName,
      latest.language,
      buffer
    );

    // Use QDoc's map for the latest score per key.
    const map = this.qdoc.getMap('strategies');
    map.set(key, JSON.stringify(score));
  }
}
