// @a0n/.ignored_gnosis/src/polyglot-optimizer.ts
/**
 * Autonomous polyglot topology optimizer.
 *
 * Observes execution times per node and automatically optimizes the topology:
 * - Slow nodes → RACE with a faster language implementation
 * - Independent nodes → FORK for parallel execution
 * - Repeated inputs → cache FOLD insertion
 * - Failing nodes → fallback RACE with alternative language
 *
 * Every optimization goes through Betty verification first.
 * The optimizer can never introduce a deadlock or resource leak.
 */

import type { GraphAST, ASTNode, ASTEdge } from './betty/compiler';

/**
 * Per-node execution profile, accumulated across invocations.
 */
export interface NodeProfile {
  nodeId: string;
  language: string;
  functionName: string;
  /** Number of times this node has been executed. */
  executions: number;
  /** Total execution time across all invocations (ms). */
  totalMs: number;
  /** Minimum execution time observed (ms). */
  minMs: number;
  /** Maximum execution time observed (ms). */
  maxMs: number;
  /** Exponentially smoothed execution time (ms). */
  smoothedMs: number;
  /** Number of errors. */
  errors: number;
  /** Number of cache hits (if caching is enabled). */
  cacheHits: number;
}

/**
 * Optimization suggestion from the analyzer.
 */
export interface OptimizationSuggestion {
  kind: 'race' | 'parallelize' | 'cache' | 'fallback';
  /** Node(s) this suggestion applies to. */
  targetNodes: string[];
  /** Human-readable description. */
  description: string;
  /** Estimated improvement factor (1.0 = no change, 2.0 = 2x faster). */
  estimatedSpeedup: number;
  /** The transformation to apply (new edges/nodes to add). */
  transform: TopologyTransform;
}

/**
 * A concrete transformation to apply to the topology.
 */
export interface TopologyTransform {
  /** Nodes to add. */
  addNodes: Array<{
    id: string;
    labels: string[];
    properties: Record<string, string>;
  }>;
  /** Edges to add. */
  addEdges: Array<{
    sourceIds: string[];
    targetIds: string[];
    type: string;
    properties: Record<string, string>;
  }>;
  /** Edges to remove (matched by source+target+type). */
  removeEdges: Array<{
    sourceIds: string[];
    targetIds: string[];
    type: string;
  }>;
}

/**
 * Accumulated execution profiles for a topology.
 */
export class TopologyProfiler {
  private profiles = new Map<string, NodeProfile>();

  /**
   * Record an execution observation for a node.
   */
  record(
    nodeId: string,
    language: string,
    functionName: string,
    durationMs: number,
    error: boolean = false,
    cacheHit: boolean = false
  ): void {
    const existing = this.profiles.get(nodeId);
    if (existing) {
      existing.executions += 1;
      existing.totalMs += durationMs;
      existing.minMs = Math.min(existing.minMs, durationMs);
      existing.maxMs = Math.max(existing.maxMs, durationMs);
      existing.smoothedMs = existing.smoothedMs * 0.7 + durationMs * 0.3;
      if (error) existing.errors += 1;
      if (cacheHit) existing.cacheHits += 1;
    } else {
      this.profiles.set(nodeId, {
        nodeId,
        language,
        functionName,
        executions: 1,
        totalMs: durationMs,
        minMs: durationMs,
        maxMs: durationMs,
        smoothedMs: durationMs,
        errors: error ? 1 : 0,
        cacheHits: cacheHit ? 1 : 0,
      });
    }
  }

  /**
   * Get the profile for a node.
   */
  get(nodeId: string): NodeProfile | undefined {
    return this.profiles.get(nodeId);
  }

  /**
   * Get all profiles.
   */
  all(): NodeProfile[] {
    return [...this.profiles.values()];
  }

  /**
   * Get the mean execution time for a node.
   */
  meanMs(nodeId: string): number {
    const p = this.profiles.get(nodeId);
    if (!p || p.executions === 0) return 0;
    return p.totalMs / p.executions;
  }

  /**
   * Get the error rate for a node (0.0 to 1.0).
   */
  errorRate(nodeId: string): number {
    const p = this.profiles.get(nodeId);
    if (!p || p.executions === 0) return 0;
    return p.errors / p.executions;
  }
}

/**
 * Analyze a topology's execution profiles and suggest optimizations.
 */
export function analyzeForOptimizations(
  ast: GraphAST,
  profiler: TopologyProfiler,
  options: {
    /** Threshold in ms above which a node is considered "slow". */
    slowThresholdMs?: number;
    /** Error rate threshold above which a fallback is suggested. */
    errorRateThreshold?: number;
    /** Minimum executions before suggesting optimizations. */
    minExecutions?: number;
  } = {}
): OptimizationSuggestion[] {
  const slowThreshold = options.slowThresholdMs ?? 500;
  const errorThreshold = options.errorRateThreshold ?? 0.1;
  const minExec = options.minExecutions ?? 5;
  const suggestions: OptimizationSuggestion[] = [];

  const profiles = profiler.all().filter((p) => p.executions >= minExec);

  // 1. Slow nodes → suggest RACE with faster language.
  for (const profile of profiles) {
    if (profile.smoothedMs > slowThreshold) {
      const fasterLang = suggestFasterLanguage(
        profile.language,
        profile.functionName
      );
      if (fasterLang) {
        const raceNodeId = `${profile.nodeId}_race_${fasterLang}`;
        suggestions.push({
          kind: 'race',
          targetNodes: [profile.nodeId],
          description:
            `Node '${profile.nodeId}' (${
              profile.language
            }) averages ${profile.smoothedMs.toFixed(0)}ms. ` +
            `RACE with ${fasterLang} implementation for first-to-finish.`,
          estimatedSpeedup: estimateLanguageSpeedup(
            profile.language,
            fasterLang
          ),
          transform: {
            addNodes: [
              {
                id: raceNodeId,
                labels: ['PolyglotBridgeCall'],
                properties: {
                  language: fasterLang,
                  callee: profile.functionName,
                  file: `${profile.functionName}.${
                    fasterLang === 'rust'
                      ? 'rs'
                      : fasterLang === 'go'
                      ? 'go'
                      : fasterLang === 'c'
                      ? 'c'
                      : 'mjs'
                  }`,
                },
              },
            ],
            addEdges: [
              // The RACE edge: run both in parallel, first-to-finish wins.
              {
                sourceIds: [profile.nodeId],
                targetIds: [raceNodeId],
                type: 'RACE',
                properties: { optimization: 'auto-race' },
              },
            ],
            removeEdges: [],
          },
        });
      }
    }
  }

  // 2. High error rate → suggest fallback RACE.
  for (const profile of profiles) {
    const rate = profiler.errorRate(profile.nodeId);
    if (rate > errorThreshold) {
      const fallbackLang =
        profile.language === 'python' ? 'javascript' : 'python';
      suggestions.push({
        kind: 'fallback',
        targetNodes: [profile.nodeId],
        description:
          `Node '${profile.nodeId}' (${profile.language}) has ${(
            rate * 100
          ).toFixed(0)}% error rate. ` + `Add ${fallbackLang} fallback.`,
        estimatedSpeedup: 1.0 / (1.0 - rate), // Improvement = inverse of failure rate
        transform: {
          addNodes: [
            {
              id: `${profile.nodeId}_fallback`,
              labels: ['PolyglotBridgeCall'],
              properties: {
                language: fallbackLang,
                callee: profile.functionName,
              },
            },
          ],
          addEdges: [
            {
              sourceIds: [profile.nodeId],
              targetIds: [`${profile.nodeId}_fallback`],
              type: 'RACE',
              properties: { optimization: 'auto-fallback' },
            },
          ],
          removeEdges: [],
        },
      });
    }
  }

  // 3. Independent sequential nodes → suggest FORK for parallelization.
  const sequentialPairs = findParallelizableSequences(ast, profiles);
  for (const pair of sequentialPairs) {
    const combinedMs = pair.reduce((sum, p) => sum + p.smoothedMs, 0);
    const parallelMs = Math.max(...pair.map((p) => p.smoothedMs));
    if (combinedMs > parallelMs * 1.3) {
      // Only suggest if >30% improvement
      suggestions.push({
        kind: 'parallelize',
        targetNodes: pair.map((p) => p.nodeId),
        description:
          `Nodes ${pair
            .map((p) => `'${p.nodeId}'`)
            .join(', ')} are independent. ` +
          `FORK for parallel execution: ${combinedMs.toFixed(
            0
          )}ms → ~${parallelMs.toFixed(0)}ms.`,
        estimatedSpeedup: combinedMs / parallelMs,
        transform: {
          addNodes: [],
          addEdges: [],
          removeEdges: [],
        },
      });
    }
  }

  return suggestions.sort((a, b) => b.estimatedSpeedup - a.estimatedSpeedup);
}

/**
 * Find sequences of nodes that could be parallelized (no data dependency).
 */
function findParallelizableSequences(
  ast: GraphAST,
  profiles: NodeProfile[]
): NodeProfile[][] {
  // Build forward adjacency.
  const forward = new Map<string, Set<string>>();
  for (const edge of ast.edges) {
    if (edge.type !== 'PROCESS') continue;
    for (const src of edge.sourceIds) {
      if (!forward.has(src.trim())) forward.set(src.trim(), new Set());
      for (const tgt of edge.targetIds) {
        forward.get(src.trim())!.add(tgt.trim());
      }
    }
  }

  // Find nodes at the same depth (no dependency between them).
  const profileMap = new Map(profiles.map((p) => [p.nodeId, p]));
  const pairs: NodeProfile[][] = [];

  // Check all FORK edges -- their targets are already parallel.
  for (const edge of ast.edges) {
    if (edge.type === 'FORK' && edge.targetIds.length >= 2) {
      const group = edge.targetIds
        .map((id) => profileMap.get(id.trim()))
        .filter((p): p is NodeProfile => p !== undefined);
      if (group.length >= 2) {
        pairs.push(group);
      }
    }
  }

  return pairs;
}

/**
 * Suggest a faster language for a function based on its characteristics.
 */
function suggestFasterLanguage(
  currentLanguage: string,
  _functionName: string
): string | null {
  // Simple heuristic: compiled languages are generally faster for compute.
  const speedTier: Record<string, number> = {
    c: 1,
    rust: 1,
    cpp: 1,
    go: 2,
    java: 3,
    kotlin: 3,
    scala: 3,
    swift: 2,
    javascript: 4,
    typescript: 4,
    python: 5,
    ruby: 5,
    lua: 5,
    php: 5,
    elixir: 4,
  };

  const currentTier = speedTier[currentLanguage] ?? 3;
  if (currentTier <= 2) return null; // Already fast.

  // Suggest one tier faster.
  if (currentTier >= 5) return 'go'; // Python/Ruby → Go
  if (currentTier >= 4) return 'rust'; // JS/Elixir → Rust
  return 'rust'; // JVM → Rust
}

/**
 * Estimate the speedup factor from switching languages.
 */
function estimateLanguageSpeedup(from: string, to: string): number {
  const benchmarks: Record<string, number> = {
    c: 1.0,
    rust: 1.1,
    cpp: 1.0,
    go: 2.5,
    java: 3.0,
    swift: 2.0,
    javascript: 5.0,
    typescript: 5.0,
    python: 30.0,
    ruby: 25.0,
    lua: 15.0,
  };

  const fromCost = benchmarks[from] ?? 5.0;
  const toCost = benchmarks[to] ?? 5.0;
  return fromCost / toCost;
}
