// @a0n/.ignored_gnosis/src/polyglot-deploy.ts
/**
 * gnode deploy -- package a polyglot topology and deploy to x-gnosis.
 *
 * Takes a .gg topology + source files, bundles them, and registers
 * with an x-gnosis instance's /.aeon/polyglot/execute endpoint.
 */

import * as fs from 'node:fs';
import * as path from 'node:path';
import { analyzePolyglotSourceForExecution } from './polyglot-bridge';
import { compose, extractFunctions } from './polyglot-compose';

/**
 * Deploy configuration.
 */
export interface DeployConfig {
  /** Source files to deploy. */
  files: string[];
  /** x-gnosis target URL (e.g., https://mysite.example.com). */
  targetUrl: string;
  /** API key for authentication (optional). */
  apiKey?: string;
  /** Function to execute as the entry point. */
  entryFunction?: string;
}

/**
 * Deploy result.
 */
export interface DeployResult {
  /** Whether the deploy succeeded. */
  success: boolean;
  /** The deployed artifact ID. */
  artifactId?: string;
  /** The endpoint URL for execution. */
  executeUrl?: string;
  /** Error message if failed. */
  error?: string;
  /** Deployed topology. */
  topology?: string;
  /** Languages involved. */
  languages?: string[];
}

/**
 * Bundle a set of source files into a deployable package.
 */
export async function bundleForDeploy(files: string[]): Promise<{
  topology: string;
  sources: Array<{ filePath: string; language: string; content: string }>;
  languages: string[];
  functionCount: number;
}> {
  // Analyze all files.
  const analyses: Array<
    Awaited<ReturnType<typeof analyzePolyglotSourceForExecution>>
  > = [];
  for (const file of files) {
    analyses.push(await analyzePolyglotSourceForExecution(file));
  }

  // Compose into a single topology.
  const composed = compose(analyses, files);

  // Collect source files.
  const sources = files.map((filePath, i) => ({
    filePath,
    language: analyses[i]!.language,
    content: fs.readFileSync(filePath, 'utf8'),
  }));

  return {
    topology: composed.ggSource,
    sources,
    languages: composed.languages,
    functionCount: composed.functions.length,
  };
}

/**
 * Deploy a polyglot topology to an x-gnosis instance.
 */
export async function deploy(config: DeployConfig): Promise<DeployResult> {
  try {
    const bundle = await bundleForDeploy(config.files);

    // POST to x-gnosis polyglot execute endpoint.
    const url = `${config.targetUrl.replace(
      /\/+$/,
      ''
    )}/.aeon/polyglot/execute`;

    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    };
    if (config.apiKey) {
      headers['Authorization'] = `Bearer ${config.apiKey}`;
    }

    const body = JSON.stringify({
      filePath: config.files[0],
      functionName: config.entryFunction,
      topology: bundle.topology,
      sources: bundle.sources,
    });

    const response = await fetch(url, {
      method: 'POST',
      headers,
      body,
    });

    if (!response.ok) {
      const text = await response.text();
      return {
        success: false,
        error: `Deploy failed (${response.status}): ${text}`,
        topology: bundle.topology,
        languages: bundle.languages,
      };
    }

    const result = (await response.json()) as { artifactId?: string };

    return {
      success: true,
      artifactId: result.artifactId,
      executeUrl: url,
      topology: bundle.topology,
      languages: bundle.languages,
    };
  } catch (err: unknown) {
    return {
      success: false,
      error: err instanceof Error ? err.message : String(err),
    };
  }
}
