// @a0n/.ignored_gnosis/src/prism-attestation.ts
/**
 * Prism Attestation — cryptographic chain-of-custody over a prism path.
 *
 * A message traverses a prism path (sequence of AeonPIDs). At each hop,
 * the receiving PID:
 *   1. Verifies the previous hop's signature
 *   2. Optionally transforms the payload (f_i(x_{i-1}) = x_i)
 *   3. Signs the transformed payload + previous chain + its own timestamp + phase
 *   4. Forwards to the next hop
 *
 * The final target receives a message with N signatures from N named PIDs
 * in a provable order, with provable timestamps and active-phase bindings.
 *
 * This composes with bitwise primitives:
 *   - Ed25519 signing (bitwise/wallet): each hop signs with its DID
 *   - FRF hashing (bitwise/cache-fp64): chain links
 *   - UCAN (bitwise/knotchain-ucan): capability scoping for what hops may attest
 *
 * The attestation layer is transport-agnostic -- use DashRelay, the
 * knotchain fabric, or any other delivery mechanism. What's provable
 * is the *chain*, not the *transport*.
 *
 * @packageDocumentation
 */

import type { PrismPath, PrismHop } from './temporal-prism';

// ─── Types ──────────────────────────────────────────────────────────

/**
 * A minimal signer interface; implementations wrap bitwise's Ed25519 wallet
 * or any other key material. The attestation layer doesn't care how the
 * signature is produced, only that it's verifiable by DID lookup.
 */
export interface AttestationSigner {
  readonly did: string;
  sign(message: Uint8Array): Promise<Uint8Array>;
}

/**
 * A minimal verifier: given a DID and a (message, signature) pair,
 * report whether the signature is valid.
 */
export interface AttestationVerifier {
  verify(did: string, message: Uint8Array, signature: Uint8Array): Promise<boolean>;
}

/**
 * One hop's contribution to the attestation chain.
 */
export interface PrismAttestationHop {
  /** Zero-indexed position in the chain. */
  readonly hopIndex: number;
  /** DID of the attester. */
  readonly attesterDid: string;
  /** PID id (atlas-scope identifier for the node). */
  readonly pidId: string;
  /** Timestamp of attestation (picolorenzos). */
  readonly attestedAtPLo: number;
  /** Solar phase at attestation time ('rising' | 'peak' | 'falling' | 'night'). */
  readonly attestedPhase: 'rising' | 'peak' | 'falling' | 'night';
  /** Hash of the payload seen at this hop (FRF hex). */
  readonly payloadHash: string;
  /** Hash of the previous hop's signature (FRF hex), or null for first hop. */
  readonly previousSignatureHash: string | null;
  /** Ed25519 signature over (payloadHash || previousSignatureHash || attestedAtPLo || attestedPhase). */
  readonly signature: Uint8Array;
  /** Optional: if this hop transformed the payload, include the output hash. */
  readonly transformedHash?: string;
  /** Optional: human-readable reason/action note. */
  readonly note?: string;
}

/**
 * A full attestation chain: the prism path plus each hop's signed attestation.
 */
export interface PrismAttestationChain {
  readonly source: 'GnosisPrismAttestationChain';
  readonly chainId: string;
  readonly originDid: string;
  readonly targetDid: string;
  readonly path: PrismPath;
  readonly hops: readonly PrismAttestationHop[];
  /** The initial payload hash (input to hop 0). */
  readonly initialPayloadHash: string;
  /** The final payload hash after all transformations. */
  readonly finalPayloadHash: string;
  readonly createdAtPLo: number;
}

/**
 * Verification result: per-hop and overall.
 */
export interface AttestationChainVerification {
  readonly source: 'GnosisAttestationChainVerification';
  readonly chainId: string;
  readonly valid: boolean;
  readonly hopResults: readonly {
    readonly hopIndex: number;
    readonly signatureValid: boolean;
    readonly chainLinkValid: boolean;
    readonly phaseConsistent: boolean;
  }[];
  readonly totalHops: number;
  readonly validHops: number;
  readonly invalidHops: number;
  readonly reasons: readonly string[];
}

// ─── Canonical Serialization ────────────────────────────────────────

/**
 * Canonical byte encoding of the attested value.
 *
 * Format: UTF-8 of "payloadHash|prevSigHash|pLo|phase"
 * where prevSigHash is the literal string "null" for the first hop.
 *
 * This is the byte string every hop signs. Any discrepancy in encoding
 * between signer and verifier will fail verification.
 */
export function canonicalAttestationBytes(
  payloadHash: string,
  previousSignatureHash: string | null,
  attestedAtPLo: number,
  attestedPhase: string
): Uint8Array {
  const canonical = [
    payloadHash,
    previousSignatureHash ?? 'null',
    attestedAtPLo.toString(),
    attestedPhase,
  ].join('|');
  return new TextEncoder().encode(canonical);
}

// ─── Chain Construction ─────────────────────────────────────────────

/**
 * FRF hash wrapper -- delegates to bitwise when available, falls back
 * to a simple SHA-256 for environments without bitwise loaded.
 *
 * Accepts a hash function parameter so the caller can inject bitwise's
 * cacheFingerprint64HexFrfSync at integration time without coupling
 * this module to bitwise directly.
 */
export type HashFn = (data: Uint8Array | string) => string;

export interface ChainBuildContext {
  readonly hashFn: HashFn;
  readonly verifier: AttestationVerifier;
}

/**
 * Input to attesting a new hop.
 */
export interface HopAttestationInput {
  readonly chain: PrismAttestationChain | null; // null for first hop
  readonly signer: AttestationSigner;
  readonly pidId: string;
  readonly attestedAtPLo: number;
  readonly attestedPhase: 'rising' | 'peak' | 'falling' | 'night';
  readonly incomingPayloadHash: string;
  readonly transformedHash?: string;
  readonly note?: string;
  readonly context: ChainBuildContext;
}

/**
 * Initialize a new attestation chain. Typically called at the origin.
 * Returns a chain with zero hops -- the first hop is added via attestHop.
 */
export function initAttestationChain(input: {
  readonly chainId: string;
  readonly originDid: string;
  readonly targetDid: string;
  readonly path: PrismPath;
  readonly initialPayloadHash: string;
  readonly createdAtPLo: number;
}): PrismAttestationChain {
  return {
    source: 'GnosisPrismAttestationChain',
    chainId: input.chainId,
    originDid: input.originDid,
    targetDid: input.targetDid,
    path: input.path,
    hops: [],
    initialPayloadHash: input.initialPayloadHash,
    finalPayloadHash: input.initialPayloadHash,
    createdAtPLo: input.createdAtPLo,
  };
}

/**
 * Attest a new hop on the chain.
 *
 * The signer computes a signature over the canonical attestation bytes
 * (payload hash + previous signature hash + pLo + phase), and the hop
 * is appended to the chain. The hash of THIS hop's signature becomes
 * the previousSignatureHash for the next hop.
 */
export async function attestHop(
  input: HopAttestationInput
): Promise<PrismAttestationChain> {
  const { chain, signer, pidId, attestedAtPLo, attestedPhase,
          incomingPayloadHash, transformedHash, note, context } = input;

  const previousSignatureHash = chain && chain.hops.length > 0
    ? context.hashFn(chain.hops[chain.hops.length - 1].signature)
    : null;

  const payloadHashForSigning = transformedHash ?? incomingPayloadHash;

  const canonical = canonicalAttestationBytes(
    payloadHashForSigning,
    previousSignatureHash,
    attestedAtPLo,
    attestedPhase
  );

  const signature = await signer.sign(canonical);

  const newHop: PrismAttestationHop = {
    hopIndex: chain ? chain.hops.length : 0,
    attesterDid: signer.did,
    pidId,
    attestedAtPLo,
    attestedPhase,
    payloadHash: payloadHashForSigning,
    previousSignatureHash,
    signature,
    ...(transformedHash !== undefined ? { transformedHash } : {}),
    ...(note !== undefined ? { note } : {}),
  };

  if (!chain) {
    // Initialize a fresh chain. Caller didn't call initAttestationChain.
    throw new Error('attestHop requires a chain; call initAttestationChain first');
  }

  return {
    ...chain,
    hops: [...chain.hops, newHop],
    finalPayloadHash: payloadHashForSigning,
  };
}

// ─── Verification ───────────────────────────────────────────────────

/**
 * Verify an attestation chain.
 *
 * Checks:
 *   1. Each hop's signature is valid (Ed25519 over canonical bytes)
 *   2. Each hop's previousSignatureHash matches hash of prior hop's signature
 *   3. Hop indices are sequential (0, 1, 2, ...)
 *   4. Attestation timestamps are monotonically increasing
 *   5. First hop's previousSignatureHash is null
 *
 * Does NOT verify: transport, actual payload content (only hash), or
 * whether the hops match the path (that's a separate check).
 */
export async function verifyAttestationChain(
  chain: PrismAttestationChain,
  context: ChainBuildContext
): Promise<AttestationChainVerification> {
  const reasons: string[] = [];
  // Local accumulator is mutable; the return type re-narrows it to readonly.
  const hopResults: Array<AttestationChainVerification['hopResults'][number]> = [];

  let lastAttestedAtPLo = -Infinity;
  let lastSigHash: string | null = null;

  for (let i = 0; i < chain.hops.length; i++) {
    const hop = chain.hops[i];

    // Check hop index is sequential
    if (hop.hopIndex !== i) {
      reasons.push(`Hop ${i} has incorrect hopIndex ${hop.hopIndex}`);
    }

    // Check previousSignatureHash matches
    const expectedPrevHash = i === 0 ? null : lastSigHash;
    const chainLinkValid = hop.previousSignatureHash === expectedPrevHash;
    if (!chainLinkValid) {
      reasons.push(`Hop ${i} previousSignatureHash does not match chain`);
    }

    // Check timestamps monotonic
    const phaseConsistent = hop.attestedAtPLo >= lastAttestedAtPLo;
    if (!phaseConsistent) {
      reasons.push(`Hop ${i} timestamp ${hop.attestedAtPLo} is before previous ${lastAttestedAtPLo}`);
    }

    // Verify signature
    const canonical = canonicalAttestationBytes(
      hop.payloadHash,
      hop.previousSignatureHash,
      hop.attestedAtPLo,
      hop.attestedPhase
    );
    let signatureValid = false;
    try {
      signatureValid = await context.verifier.verify(hop.attesterDid, canonical, hop.signature);
    } catch (err) {
      reasons.push(`Hop ${i} verifier threw: ${String(err)}`);
    }
    if (!signatureValid) {
      reasons.push(`Hop ${i} signature invalid for DID ${hop.attesterDid}`);
    }

    hopResults.push({
      hopIndex: i,
      signatureValid,
      chainLinkValid,
      phaseConsistent,
    });

    lastAttestedAtPLo = hop.attestedAtPLo;
    lastSigHash = context.hashFn(hop.signature);
  }

  const validHops = hopResults.filter(r =>
    r.signatureValid && r.chainLinkValid && r.phaseConsistent
  ).length;

  return {
    source: 'GnosisAttestationChainVerification',
    chainId: chain.chainId,
    valid: validHops === chain.hops.length && chain.hops.length > 0,
    hopResults,
    totalHops: chain.hops.length,
    validHops,
    invalidHops: chain.hops.length - validHops,
    reasons,
  };
}

// ─── Chain Analysis ─────────────────────────────────────────────────

/**
 * A summary of who signed a chain and when.
 */
export interface AttestationChainSummary {
  readonly source: 'GnosisAttestationChainSummary';
  readonly chainId: string;
  readonly hopCount: number;
  readonly distinctAttesters: number;
  readonly firstAttestationPLo: number;
  readonly lastAttestationPLo: number;
  readonly elapsedPLo: number;
  readonly payloadTransformed: boolean;
  readonly phaseDistribution: Record<string, number>;
}

/**
 * Handles the gnosis summarize Attestation Chain workflow.
 */
export function summarizeAttestationChain(
  chain: PrismAttestationChain
): AttestationChainSummary {
  const attesters = new Set<string>();
  const phaseDistribution: Record<string, number> = {};
  for (const hop of chain.hops) {
    attesters.add(hop.attesterDid);
    phaseDistribution[hop.attestedPhase] = (phaseDistribution[hop.attestedPhase] ?? 0) + 1;
  }

  const first = chain.hops[0]?.attestedAtPLo ?? 0;
  const last = chain.hops[chain.hops.length - 1]?.attestedAtPLo ?? 0;

  return {
    source: 'GnosisAttestationChainSummary',
    chainId: chain.chainId,
    hopCount: chain.hops.length,
    distinctAttesters: attesters.size,
    firstAttestationPLo: first,
    lastAttestationPLo: last,
    elapsedPLo: Math.max(0, last - first),
    payloadTransformed: chain.initialPayloadHash !== chain.finalPayloadHash,
    phaseDistribution,
  };
}

/**
 * Does the chain's hop sequence match the prism path's hop sequence?
 *
 * Checks that chain.hops[i].pidId matches path.hops[i].toPidId
 * (and hop 0 corresponds to origin → path.hops[0].toPidId).
 */
export function chainMatchesPath(
  chain: PrismAttestationChain,
  path: PrismPath
): { readonly matches: boolean; readonly mismatches: readonly number[] } {
  const mismatches: number[] = [];
  const expectedPids = path.hops.map(h => h.toPidId);
  const actualPids = chain.hops.map(h => h.pidId);

  const minLen = Math.min(expectedPids.length, actualPids.length);
  for (let i = 0; i < minLen; i++) {
    if (expectedPids[i] !== actualPids[i]) {
      mismatches.push(i);
    }
  }
  if (expectedPids.length !== actualPids.length) {
    mismatches.push(-1); // length mismatch sentinel
  }

  return {
    matches: mismatches.length === 0,
    mismatches,
  };
}
