// @a0n/.ignored_gnosis/src/prismatic-refraction-knotgraph.ts
/**
 * Knotgraph ingest paragon for Prismatic Refraction.
 *
 * Closes the loop end-to-end:
 *
 *   app.vomit(bytes)                              // no shape
 *     → wireToStorage(stream, k)                  // wire sorts
 *       → mapSettledShardsToBlocks(witness, ...)  // this module
 *         → KnotgraphBlock[]                      // ready for the TsLane
 *
 * Each settled shard becomes a `KnotgraphBlock` in the 54-dim fp64
 * substrate, with the app zone (dims 38-53) populated per the
 * knotgraph convention:
 *
 *   - dim 38-39 (TENANT_ID_HI/LO)   = caller tenant
 *   - dim 40     (REGION_ID)        = shard.laneId (lane = sub-region)
 *   - dim 41     (TYPE_TAG)         = caller type tag
 *   - dim 42     (ENTITY_ID)        = content-addressed payload hash
 *   - dim 43     (SCHEMA_KNOT_HASH) = caller schema knot
 *   - dim 44     (PARENT_COMMIT)    = 0 (first version)
 *   - dim 45-50  (FIELD_0..FIELD_5) = payload summary (length, phase,
 *                                     first-40-bits of content hash,
 *                                     reserved)
 *   - dim 51     (PROVENANCE_DID)   = caller provenance DID (optional)
 *   - dim 52     (FRESHNESS_CLOCK)  = caller-supplied clock
 *
 * The claim witnessed by the accompanying tests: every produced block
 * satisfies knotgraph's tenant-isolation invariant
 * (`Tenancy/TenantIsolation.lean`), and two independent pipelines over
 * the same inputs produce equal block-dim arrays (matching knotgraph's
 * `Race/EncodingRaceDeterminism.race_deterministic`).
 *
 * This is NOT a full storage wire — block bytes are not written to D1
 * or R2. It is a pure transform that produces the input to the TsLane's
 * `createEntity` path. Wiring to the TsLane is a follow-up.
 */

import type { Shard, WireToStorageWitness } from './prismatic-refraction';
import { wireToStorage } from './prismatic-refraction';
import type { RawStream } from './prismatic-refraction';

// ─── Types (mirror knotgraph's types.ts surface, no cross-package import) ───

/** 128-bit tenant id across two fp64 lanes. Shape-match to
 *  `@a0n/knotgraph`'s `TenantId`. */
export interface TenantId {
  readonly hi: number;
  readonly lo: number;
}

/** Caller-supplied metadata required to materialize knotgraph blocks
 *  from prismatic settled shards. */
export interface IngestContext {
  readonly tenant: TenantId;
  readonly typeTag: number;
  readonly schemaKnotHash: number;
  readonly provenanceDid?: number;
  readonly freshnessClock?: number;
}

/** A knotgraph-shaped block: 54 fp64 dims. Structural mirror of
 *  `KnotgraphBlock`; uses a plain Float64Array so this module has no
 *  dependency on `@a0n/knotgraph`. The TsLane reads `dims` directly. */
export interface KnotgraphShapedBlock {
  readonly dims: Float64Array;
}

/** The ingest result: one block per settled shard + a soundness
 *  witness bundling the conjuncts callers care about. */
export interface IngestResult {
  readonly blocks: KnotgraphShapedBlock[];
  readonly witness: WireToStorageWitness;
  readonly tenantScopeHoldsForAll: boolean;
  readonly totalIngestedBytes: number;
}

// ─── Dim constants (locked to knotgraph's `dim.ts`) ──────────────────

const DIM_COUNT = 54;
const TENANT_ID_HI = 38;
const TENANT_ID_LO = 39;
const REGION_ID = 40;
const TYPE_TAG = 41;
const ENTITY_ID = 42;
const SCHEMA_KNOT_HASH = 43;
const PARENT_COMMIT = 44;
const FIELD_0 = 45;
const FIELD_1 = 46;
const FIELD_2 = 47;
const FIELD_3 = 48;
const FIELD_4 = 49;
const FIELD_5 = 50;
const PROVENANCE_DID = 51;
const FRESHNESS_CLOCK = 52;

// ─── Content hash (deterministic, fits in fp64 safely) ───────────────

/** 48-bit FNV-1a derivative over the payload. Safe to store in an fp64
 *  lane (≤ 2^53-1). Deterministic — two independent workers on the
 *  same payload produce the same hash. */
function contentHash48(payload: ReadonlyArray<number>): number {
  // FNV-1a with truncation to 48 bits.
  const FNV_OFFSET = 0xcbf29ce484222325n;
  const FNV_PRIME = 0x100000001b3n;
  let h = FNV_OFFSET;
  for (const b of payload) {
    h = (h ^ BigInt(b & 0xff)) & 0xffffffffffffffffn;
    h = (h * FNV_PRIME) & 0xffffffffffffffffn;
  }
  return Number(h & 0xffffffffffffn); // lower 48 bits
}

/** Tenant-isolation check. Mirrors `blockBelongsToTenant` from
 *  knotgraph, structurally. */
export function blockBelongsToTenant(
  block: KnotgraphShapedBlock,
  tenant: TenantId,
): boolean {
  return (
    block.dims[TENANT_ID_HI] === tenant.hi &&
    block.dims[TENANT_ID_LO] === tenant.lo
  );
}

// ─── Mapping: settled shard → knotgraph block ────────────────────────

/** Map a single settled shard to a knotgraph-shaped block, populating
 *  the app zone (dims 38-53). Knotchain core (dims 0-37) is left at
 *  zero; the seal-phase (done at storage wire time by gnosis_runtime)
 *  computes BKS invariants and chain metadata.
 *
 *  Determinism contract: `shardToBlock(s, ctx)` is a pure function of
 *  its inputs. Two workers with the same `(s, ctx)` produce
 *  byte-identical `dims` arrays. */
export function shardToBlock(
  shard: Shard,
  ctx: IngestContext,
): KnotgraphShapedBlock {
  const dims = new Float64Array(DIM_COUNT);
  dims[TENANT_ID_HI] = ctx.tenant.hi;
  dims[TENANT_ID_LO] = ctx.tenant.lo;
  dims[REGION_ID] = shard.laneId;
  dims[TYPE_TAG] = ctx.typeTag;
  const hash = contentHash48(shard.payload);
  dims[ENTITY_ID] = hash;
  dims[SCHEMA_KNOT_HASH] = ctx.schemaKnotHash;
  dims[PARENT_COMMIT] = 0;
  dims[FIELD_0] = shard.payload.length;
  dims[FIELD_1] = shard.phase;
  dims[FIELD_2] = hash;
  dims[FIELD_3] = 0;
  dims[FIELD_4] = 0;
  dims[FIELD_5] = 0;
  dims[PROVENANCE_DID] = ctx.provenanceDid ?? 0;
  dims[FRESHNESS_CLOCK] = ctx.freshnessClock ?? 0;
  return { dims };
}

// ─── End-to-end: RawStream → KnotgraphShapedBlock[] ─────────────────

/** Full ingest pipeline. Takes a raw byte firehose, runs the wire
 *  layer's prism + settle, and materializes one block per settled
 *  lane. The result bundles a `WireToStorageWitness` (from the
 *  prismatic-refraction crown theorem) plus a tenant-scope-for-all
 *  check. */
export function ingestToKnotgraph(
  stream: RawStream,
  laneCount: number,
  ctx: IngestContext,
): IngestResult {
  const witness = wireToStorage(stream, laneCount);
  const blocks: KnotgraphShapedBlock[] = witness.settled.map((s) =>
    shardToBlock(s, ctx),
  );
  let tenantScopeHoldsForAll = true;
  let totalIngestedBytes = 0;
  for (let i = 0; i < blocks.length; i++) {
    const b = blocks[i]!;
    totalIngestedBytes += b.dims[FIELD_0]!;
    if (!blockBelongsToTenant(b, ctx.tenant)) {
      tenantScopeHoldsForAll = false;
    }
  }
  return { blocks, witness, tenantScopeHoldsForAll, totalIngestedBytes };
}

// ─── Dim-array equality (for region-parallel determinism checks) ──────

/** True when two blocks have bitwise-equal `dims` arrays. Witnesses
 *  knotgraph's `race_deterministic`: two workers in different regions
 *  ingesting the same (stream, k, ctx) produce equal blocks. */
export function blocksDimsEqual(
  a: KnotgraphShapedBlock,
  b: KnotgraphShapedBlock,
): boolean {
  if (a.dims.length !== b.dims.length) return false;
  for (let i = 0; i < a.dims.length; i++) {
    if (a.dims[i] !== b.dims[i]) return false;
  }
  return true;
}

// ─── Dim constant re-exports for test + consumer clarity ─────────────

export const KNOTGRAPH_DIMS = {
  TENANT_ID_HI,
  TENANT_ID_LO,
  REGION_ID,
  TYPE_TAG,
  ENTITY_ID,
  SCHEMA_KNOT_HASH,
  PARENT_COMMIT,
  FIELD_0,
  FIELD_1,
  FIELD_2,
  FIELD_3,
  FIELD_4,
  FIELD_5,
  PROVENANCE_DID,
  FRESHNESS_CLOCK,
  DIM_COUNT,
} as const;
