// @a0n/.ignored_gnosis/src/prismatic-refraction-fast.ts
/**
 * Prismatic Refraction — typed-array fast path.
 *
 * Operational specialization of `prismatic-refraction.ts`. The slow path
 * allocates one `Shard` object per input byte and uses `BigInt` for the
 * content hash; that hits ~25 MB/s on 1 MB payloads in pure TS. This
 * module collapses lane assignment to a two-pass count-then-fill over
 * `Uint8Array` lane buffers and uses a `Math.imul`-based 32-bit hash
 * combined into a 48-bit-safe `number`.
 *
 * Semantics preserved vs. slow path:
 *   - Lane assignment: `bytes[i] % laneCount` (matches
 *     `prismRespectsLaneBound`).
 *   - Byte-order within a lane is preserved.
 *   - Total byte count across lanes equals input length.
 *   - Per-lane payload is content-addressed by a deterministic hash.
 *
 * Semantics NOT matched:
 *   - The ENTITY_ID hash is a Math.imul FNV-1a-variant, not the slow
 *     path's BigInt FNV-1a. The fast-path hash is deterministic but
 *     diverges from the slow-path value by construction.
 *
 * Formal anchor: `ForkRaceFoldTheorems.PrismaticRefraction`
 * (`prismRespectsLaneBound`, `prismPreservesByteCount`).
 */

import { KNOTGRAPH_DIMS } from './prismatic-refraction-knotgraph';

// ─── FastIngestContext (mirrors IngestContext sans object churn) ─────

/** 128-bit tenant id across two fp64 lanes. Shape-match to
 *  `prismatic-refraction-knotgraph.ts`'s `TenantId`. */
export interface FastTenantId {
  readonly hi: number;
  readonly lo: number;
}

/** Caller-supplied metadata for fast-path block materialization.
 *  Mirrors `IngestContext`. */
export interface FastIngestContext {
  readonly tenant: FastTenantId;
  readonly typeTag: number;
  readonly schemaKnotHash: number;
  readonly provenanceDid?: number;
  readonly freshnessClock?: number;
}

/** Knotgraph-shaped block — same shape as `KnotgraphShapedBlock`. */
export interface FastBlock {
  readonly dims: Float64Array;
}

// ─── Content hash (Math.imul 32-bit, combined to 48-bit-safe number) ─

const HASH_SEED_A = 0x811c9dc5 | 0;
const HASH_SEED_B = 0x1b873593 | 0;
const HASH_MULT_A = 0x01000193 | 0;
const HASH_MULT_B = 0x5bd1e995 | 0;

/** Deterministic 48-bit content hash over a Uint8Array slice.
 *  Two 32-bit FNV-1a-style halves combined as `hi * 2^16 + lo` so the
 *  result fits in fp64 without precision loss (48 bits ≤ 53).
 *  Does not match the slow-path BigInt FNV-1a value. */
function contentHash48Fast(
  buf: Uint8Array,
  start: number,
  end: number,
): number {
  let a = HASH_SEED_A;
  let b = HASH_SEED_B;
  for (let i = start; i < end; i++) {
    const x = buf[i]! | 0;
    a = Math.imul(a ^ x, HASH_MULT_A);
    b = Math.imul(b ^ x, HASH_MULT_B);
  }
  // Mix to decorrelate.
  a ^= a >>> 16;
  b ^= b >>> 13;
  // Fold into a 48-bit-safe positive fp64. Use the low 16 bits of `a`
  // as the high half, low 32 of `b` as the low half → 48 bits total.
  const hi16 = (a >>> 0) & 0xffff;
  const lo32 = b >>> 0;
  return hi16 * 0x100000000 + lo32; // fits in 2^48
}

// ─── prismFast: two-pass count-then-fill over Uint8Array lanes ───────

/** Content-addressed lane-bucketing over a `Uint8Array`. One pre-sized
 *  `Uint8Array` per lane, byte-order-preserving. Two-pass (count, fill)
 *  to avoid growth reallocations.
 *
 *  Lane assignment: `bytes[i] % laneCount`. Matches
 *  `prismRespectsLaneBound` and `prismPreservesByteCount`. */
export function prismFast(
  bytes: Uint8Array,
  laneCount: number,
): Uint8Array[] {
  if (laneCount <= 0) {
    throw new RangeError(`laneCount must be positive, got ${laneCount}`);
  }
  const n = bytes.length;

  // Pass 1: per-lane counts.
  const counts = new Int32Array(laneCount);
  if (laneCount === 1) {
    counts[0] = n;
  } else {
    for (let i = 0; i < n; i++) {
      counts[bytes[i]! % laneCount]!++;
    }
  }

  // Pre-size lane buffers.
  const lanes: Uint8Array[] = new Array(laneCount);
  for (let l = 0; l < laneCount; l++) lanes[l] = new Uint8Array(counts[l]!);

  // Pass 2: fill. Reuse `counts` as write cursors (reset to zero).
  if (laneCount === 1) {
    lanes[0]!.set(bytes);
  } else {
    const cursors = new Int32Array(laneCount);
    for (let i = 0; i < n; i++) {
      const byte = bytes[i]!;
      const lane = byte % laneCount;
      const pos = cursors[lane]!;
      lanes[lane]![pos] = byte;
      cursors[lane] = pos + 1;
    }
  }

  return lanes;
}

// ─── settleFastToBlocks: per-lane Float64Array blocks ────────────────

const {
  TENANT_ID_HI,
  TENANT_ID_LO,
  REGION_ID,
  TYPE_TAG,
  ENTITY_ID,
  SCHEMA_KNOT_HASH,
  PARENT_COMMIT,
  FIELD_0,
  FIELD_1,
  FIELD_2,
  PROVENANCE_DID,
  FRESHNESS_CLOCK,
  DIM_COUNT,
} = KNOTGRAPH_DIMS;

/** Materialize one knotgraph-shaped block per lane that has bytes.
 *  Byte-order within each lane is preserved from `prismFast`. Empty
 *  lanes are skipped to mirror the slow path's `settleByLane`
 *  semantics (only distinct observed lanes emit a shard). */
export function settleFastToBlocks(
  lanes: Uint8Array[],
  ctx: FastIngestContext,
): FastBlock[] {
  const provenance = ctx.provenanceDid ?? 0;
  const freshness = ctx.freshnessClock ?? 0;
  const blocks: FastBlock[] = [];
  for (let laneId = 0; laneId < lanes.length; laneId++) {
    const buf = lanes[laneId]!;
    if (buf.length === 0) continue;
    const dims = new Float64Array(DIM_COUNT);
    const hash = contentHash48Fast(buf, 0, buf.length);
    dims[TENANT_ID_HI] = ctx.tenant.hi;
    dims[TENANT_ID_LO] = ctx.tenant.lo;
    dims[REGION_ID] = laneId;
    dims[TYPE_TAG] = ctx.typeTag;
    dims[ENTITY_ID] = hash;
    dims[SCHEMA_KNOT_HASH] = ctx.schemaKnotHash;
    dims[PARENT_COMMIT] = 0;
    dims[FIELD_0] = buf.length;
    dims[FIELD_1] = 0; // phase — always 0 on the fast path (no refract folds yet).
    dims[FIELD_2] = hash;
    dims[PROVENANCE_DID] = provenance;
    dims[FRESHNESS_CLOCK] = freshness;
    blocks.push({ dims });
  }
  return blocks;
}

// ─── wireToStorageFast: end-to-end ───────────────────────────────────

/** End-to-end wire pipeline over a `Uint8Array` input. Returns the
 *  knotgraph-shaped blocks plus the total bytes accounted for across
 *  lanes. */
export function wireToStorageFast(
  bytes: Uint8Array,
  k: number,
  ctx: FastIngestContext,
): { blocks: FastBlock[]; totalBytes: number } {
  if (k <= 0) {
    throw new RangeError(`laneCount must be positive, got ${k}`);
  }
  if (k > 16) {
    throw new RangeError(
      `laneCount must be <= 16 to fit the knotgraph app zone (dims 38-53), got ${k}`,
    );
  }
  const lanes = prismFast(bytes, k);
  const blocks = settleFastToBlocks(lanes, ctx);
  let totalBytes = 0;
  for (let i = 0; i < lanes.length; i++) totalBytes += lanes[i]!.length;
  return { blocks, totalBytes };
}

// ─── Validators (used by bench; cheap O(n) passes) ───────────────────

/** True if every lane index is in `[0, laneCount)` by construction.
 *  This is a structural invariant of `prismFast` (lane ids are array
 *  indices below `laneCount`); the function checks array shape. */
export function lanesBounded(lanes: Uint8Array[], k: number): boolean {
  if (lanes.length !== k) return false;
  for (let i = 0; i < lanes.length; i++) {
    if (!(lanes[i] instanceof Uint8Array)) return false;
  }
  return true;
}

/** True if the total byte count across lanes matches the input length
 *  and every lane's bytes actually satisfy `byte % k === laneId`. */
export function payloadPreserved(
  lanes: Uint8Array[],
  k: number,
  inputLen: number,
): boolean {
  let total = 0;
  for (let laneId = 0; laneId < lanes.length; laneId++) {
    const buf = lanes[laneId]!;
    total += buf.length;
    for (let i = 0; i < buf.length; i++) {
      if (buf[i]! % k !== laneId) return false;
    }
  }
  return total === inputLen;
}
