// @a0n/.ignored_gnosis/src/protocol69-r1-phyle-entropy.ts
import {
  buildProtocol69ProjectionEnvelope,
  type Protocol69ProjectionEnvelope,
} from './protocol69';
import { validateProtocol69ProjectionEnvelope } from './protocol69-r1-witness';

export const R1_PHYLE_ENTROPY_SCHEMA_VERSION =
  'gnosis.protocol69.r1-phyle-entropy.v1';
export const R1_PHYLE_ENTROPY_CONTRACT_ID =
  'gnosis.protocol69.r1-phyle-entropy.cooperative-escrow.v1';
export const R1_PHYLE_ENTROPY_THEOREM_ID =
  'FoilLoserPolicy.r1_phyle_escrows_surplus_entropy_without_compute_authority';

export interface R1PhyleEntropyEnvelopeInput {
  readonly deviceId: string;
  readonly phyleId: string;
  readonly sinkProcessId: string;
  readonly surplusUnits: number;
  readonly acceptedUnits?: number;
  readonly escrowedUnits?: number;
  readonly redeliveryToken?: string;
  readonly entropyWord: number;
  readonly cadenceMs?: number;
  readonly timestampMs: number;
  readonly expiresAtMs: number;
  readonly projection?: Protocol69ProjectionEnvelope;
}

export interface R1PhyleEntropyEnvelope extends R1PhyleEntropyEnvelopeInput {
  readonly schemaVersion: typeof R1_PHYLE_ENTROPY_SCHEMA_VERSION;
  readonly contractId: typeof R1_PHYLE_ENTROPY_CONTRACT_ID;
  readonly egress: 'PhyleEntropyEscrow';
  readonly theoremId: typeof R1_PHYLE_ENTROPY_THEOREM_ID;
  readonly executable: false;
  readonly acceptedUnits: number;
  readonly escrowedUnits: number;
  readonly redeliveryToken: string;
}

function asRecord(value: unknown): Record<string, unknown> | null {
  if (!value || typeof value !== 'object' || Array.isArray(value)) {
    return null;
  }
  return value as Record<string, unknown>;
}

function normalizeNonemptyString(value: unknown, name: string): string {
  if (typeof value !== 'string') {
    throw new Error(`${name} must be a string`);
  }
  const trimmed = value.trim();
  if (!trimmed) {
    throw new Error(`${name} must be nonempty`);
  }
  return trimmed;
}

function normalizeInteger(
  value: unknown,
  name: string,
  min: number,
  max: number
): number {
  if (
    typeof value !== 'number' ||
    !Number.isFinite(value) ||
    !Number.isInteger(value) ||
    value < min ||
    value > max
  ) {
    throw new Error(`${name} must be an integer in [${min}, ${max}]`);
  }
  return value;
}

function normalizeOptionalU32(value: unknown, name: string): number | undefined {
  if (value === undefined) {
    return undefined;
  }
  return normalizeInteger(value, name, 0, 0xffff_ffff);
}

function buildRedeliveryToken(input: {
  readonly phyleId: string;
  readonly deviceId: string;
  readonly sinkProcessId: string;
  readonly timestampMs: number;
  readonly entropyWord: number;
}): string {
  return [
    'r1-phyle-redelivery',
    input.phyleId,
    input.sinkProcessId,
    input.deviceId,
    String(input.timestampMs),
    input.entropyWord.toString(16),
  ].join(':');
}

/**
 * Builds the R1 Phyle Entropy Envelope.
 */
export function buildR1PhyleEntropyEnvelope(
  input: R1PhyleEntropyEnvelopeInput
): R1PhyleEntropyEnvelope {
  const cadenceMs = normalizeOptionalU32(input.cadenceMs, 'cadenceMs');
  const projection = input.projection
    ? validateProtocol69ProjectionEnvelope(input.projection)
    : buildProtocol69ProjectionEnvelope();
  const timestampMs = normalizeInteger(
    input.timestampMs,
    'timestampMs',
    0,
    Number.MAX_SAFE_INTEGER
  );
  const expiresAtMs = normalizeInteger(
    input.expiresAtMs,
    'expiresAtMs',
    timestampMs,
    Number.MAX_SAFE_INTEGER
  );
  const surplusUnits = normalizeInteger(input.surplusUnits, 'surplusUnits', 0, 64);
  const acceptedUnits = normalizeInteger(
    input.acceptedUnits ?? surplusUnits,
    'acceptedUnits',
    0,
    surplusUnits
  );
  const escrowedUnits = normalizeInteger(
    input.escrowedUnits ?? surplusUnits - acceptedUnits,
    'escrowedUnits',
    0,
    surplusUnits
  );
  if (acceptedUnits + escrowedUnits !== surplusUnits) {
    throw new Error('acceptedUnits + escrowedUnits must equal surplusUnits');
  }
  const deviceId = normalizeNonemptyString(input.deviceId, 'deviceId');
  const phyleId = normalizeNonemptyString(input.phyleId, 'phyleId');
  const sinkProcessId = normalizeNonemptyString(
    input.sinkProcessId,
    'sinkProcessId'
  );
  const entropyWord = normalizeInteger(input.entropyWord, 'entropyWord', 0, 0xffff);

  return {
    schemaVersion: R1_PHYLE_ENTROPY_SCHEMA_VERSION,
    contractId: R1_PHYLE_ENTROPY_CONTRACT_ID,
    egress: 'PhyleEntropyEscrow',
    theoremId: R1_PHYLE_ENTROPY_THEOREM_ID,
    executable: false,
    deviceId,
    phyleId,
    sinkProcessId,
    surplusUnits,
    acceptedUnits,
    escrowedUnits,
    redeliveryToken:
      input.redeliveryToken ??
      buildRedeliveryToken({
        phyleId,
        deviceId,
        sinkProcessId,
        timestampMs,
        entropyWord,
      }),
    entropyWord,
    ...(cadenceMs === undefined ? {} : { cadenceMs }),
    timestampMs,
    expiresAtMs,
    projection,
  };
}

/**
 * Handles the gnosis validate R1 Phyle Entropy Envelope workflow.
 */
export function validateR1PhyleEntropyEnvelope(
  value: unknown
): R1PhyleEntropyEnvelope {
  const record = asRecord(value);
  if (!record) {
    throw new Error('R1 phyle entropy envelope must be an object');
  }
  if (record['schemaVersion'] !== R1_PHYLE_ENTROPY_SCHEMA_VERSION) {
    throw new Error('schemaVersion mismatch');
  }
  if (record['contractId'] !== R1_PHYLE_ENTROPY_CONTRACT_ID) {
    throw new Error('contractId mismatch');
  }
  if (record['egress'] !== 'PhyleEntropyEscrow') {
    throw new Error('egress must be PhyleEntropyEscrow');
  }
  if (record['theoremId'] !== R1_PHYLE_ENTROPY_THEOREM_ID) {
    throw new Error('theoremId mismatch');
  }
  if (record['executable'] !== false) {
    throw new Error('phyle entropy envelopes must be non-executable');
  }

  const cadenceMs = normalizeOptionalU32(record['cadenceMs'], 'cadenceMs');
  return buildR1PhyleEntropyEnvelope({
    deviceId: normalizeNonemptyString(record['deviceId'], 'deviceId'),
    phyleId: normalizeNonemptyString(record['phyleId'], 'phyleId'),
    sinkProcessId: normalizeNonemptyString(
      record['sinkProcessId'],
      'sinkProcessId'
    ),
    surplusUnits: normalizeInteger(record['surplusUnits'], 'surplusUnits', 0, 64),
    acceptedUnits: normalizeInteger(record['acceptedUnits'], 'acceptedUnits', 0, 64),
    escrowedUnits: normalizeInteger(record['escrowedUnits'], 'escrowedUnits', 0, 64),
    redeliveryToken: normalizeNonemptyString(
      record['redeliveryToken'],
      'redeliveryToken'
    ),
    entropyWord: normalizeInteger(record['entropyWord'], 'entropyWord', 0, 0xffff),
    ...(cadenceMs === undefined ? {} : { cadenceMs }),
    timestampMs: normalizeInteger(
      record['timestampMs'],
      'timestampMs',
      0,
      Number.MAX_SAFE_INTEGER
    ),
    expiresAtMs: normalizeInteger(
      record['expiresAtMs'],
      'expiresAtMs',
      0,
      Number.MAX_SAFE_INTEGER
    ),
    projection: validateProtocol69ProjectionEnvelope(record['projection']),
  });
}
