// @a0n/.ignored_gnosis/src/protocol69-r1-worksteal.ts
import {
  buildProtocol69ProjectionEnvelope,
  type Protocol69ProjectionEnvelope,
} from './protocol69';
import { validateProtocol69ProjectionEnvelope } from './protocol69-r1-witness';

export const R1_WORKSTEAL_HINT_SCHEMA_VERSION =
  'gnosis.protocol69.r1-worksteal-hint.v1';
export const R1_WORKSTEAL_HINT_CONTRACT_ID =
  'gnosis.protocol69.r1-worksteal.deterministic-fib-hint.v1';
export const R1_WORKSTEAL_HINT_THEOREM_ID =
  'FoilLoserPolicy.r1_idle_witness_may_try_deterministic_work_hint';

export const R1_COMPUTE_RECEIPT_SCHEMA_VERSION =
  'gnosis.protocol69.r1-compute-receipt.v1';
export const R1_COMPUTE_RECEIPT_CONTRACT_ID =
  'gnosis.protocol69.r1-worksteal.deterministic-fib-receipt.v1';
export const R1_COMPUTE_RECEIPT_THEOREM_ID =
  'FoilLoserPolicy.r1_compute_receipt_beams_cache_reuse_rights_not_authority';

export const R1_WORKSTEAL_CLAIM_SCHEMA_VERSION =
  'gnosis.protocol69.r1-worksteal-claim.v1';
export const R1_WORKSTEAL_CLAIM_CONTRACT_ID =
  'gnosis.protocol69.r1-worksteal.resident-claim.v1';
export const R1_WORKSTEAL_CLAIM_THEOREM_ID =
  'FoilLoserPolicy.r1_resident_worker_claims_admitted_hint';

export const R1_WORKSTEAL_QUEUE_SUMMARY_SCHEMA_VERSION =
  'gnosis.protocol69.r1-worksteal-queue-summary.v1';
export const R1_WORKSTEAL_QUEUE_SUMMARY_CONTRACT_ID =
  'gnosis.protocol69.r1-worksteal.queue-flow-summary.v1';

export const R1_WORKSTEAL_RECEIPT_SUBMISSION_SCHEMA_VERSION =
  'gnosis.protocol69.r1-worksteal-receipt-submission.v1';
export const R1_WORKSTEAL_RECEIPT_SUBMISSION_CONTRACT_ID =
  'gnosis.protocol69.r1-worksteal.receipt-submission.v1';
export const R1_WORKSTEAL_RECEIPT_SUBMISSION_THEOREM_ID =
  'FoilLoserPolicy.r1_receipt_submission_preserves_reuse_right_boundary';

export interface R1WorkStealHintInput {
  readonly hintId: string;
  readonly producerDeviceId: string;
  readonly targetPhyleId: string;
  readonly workKind: 'fibonacci' | 'qwen-knot';
  readonly n: number;
  readonly modelId?: string;
  readonly knotId?: string;
  readonly knotHash?: string;
  readonly tokenBudget?: number;
  readonly priority?: number;
  readonly ttlMs: number;
  readonly timestampMs: number;
  readonly witnessWindowHash?: string;
  readonly projection?: Protocol69ProjectionEnvelope;
}

export interface R1WorkStealHint extends R1WorkStealHintInput {
  readonly schemaVersion: typeof R1_WORKSTEAL_HINT_SCHEMA_VERSION;
  readonly contractId: typeof R1_WORKSTEAL_HINT_CONTRACT_ID;
  readonly theoremId: typeof R1_WORKSTEAL_HINT_THEOREM_ID;
  readonly executable: false;
}

export interface R1ComputeReceiptInput {
  readonly hintId: string;
  readonly workerDeviceId: string;
  readonly targetPhyleId: string;
  readonly workKind: 'fibonacci' | 'qwen-knot';
  readonly n: number;
  readonly modelId?: string;
  readonly knotId?: string;
  readonly knotHash?: string;
  readonly tokenBudget?: number;
  readonly result: string;
  readonly resultDigest: string;
  readonly computeNs: number;
  readonly validationNs: number;
  readonly timestampMs: number;
  readonly beamCacheKey: string;
  readonly witnessWindowHash?: string;
  readonly projection?: Protocol69ProjectionEnvelope;
}

export interface R1ComputeReceipt extends R1ComputeReceiptInput {
  readonly schemaVersion: typeof R1_COMPUTE_RECEIPT_SCHEMA_VERSION;
  readonly contractId: typeof R1_COMPUTE_RECEIPT_CONTRACT_ID;
  readonly theoremId: typeof R1_COMPUTE_RECEIPT_THEOREM_ID;
  readonly authority: 'reuse-right';
  readonly executable: false;
}

export interface R1WorkStealPhyleQueueSummary {
  readonly phyleId: string;
  readonly pending: number;
  readonly claimed: number;
  readonly accepted: number;
  readonly rejected: number;
  readonly completedUnits: number;
}

export interface R1WorkStealQueueSummaryInput {
  readonly pending: number;
  readonly claimed: number;
  readonly accepted: number;
  readonly rejected: number;
  readonly perPhyle: readonly R1WorkStealPhyleQueueSummary[];
  readonly timestampMs: number;
}

export interface R1WorkStealQueueSummary
  extends R1WorkStealQueueSummaryInput {
  readonly schemaVersion: typeof R1_WORKSTEAL_QUEUE_SUMMARY_SCHEMA_VERSION;
  readonly contractId: typeof R1_WORKSTEAL_QUEUE_SUMMARY_CONTRACT_ID;
}

export interface R1WorkStealClaimInput {
  readonly status: 'claimed' | 'idle';
  readonly workerDeviceId: string;
  readonly phyleId: string;
  readonly claimId?: string;
  readonly hint?: R1WorkStealHint;
  readonly summary: R1WorkStealQueueSummary;
  readonly timestampMs: number;
  readonly projection?: Protocol69ProjectionEnvelope;
}

export interface R1WorkStealClaim extends R1WorkStealClaimInput {
  readonly schemaVersion: typeof R1_WORKSTEAL_CLAIM_SCHEMA_VERSION;
  readonly contractId: typeof R1_WORKSTEAL_CLAIM_CONTRACT_ID;
  readonly theoremId: typeof R1_WORKSTEAL_CLAIM_THEOREM_ID;
  readonly executable: false;
}

export interface R1WorkStealReceiptSubmissionInput {
  readonly claimId: string;
  readonly receipt: R1ComputeReceipt;
  readonly timestampMs: number;
  readonly projection?: Protocol69ProjectionEnvelope;
}

export interface R1WorkStealReceiptSubmission
  extends R1WorkStealReceiptSubmissionInput {
  readonly schemaVersion: typeof R1_WORKSTEAL_RECEIPT_SUBMISSION_SCHEMA_VERSION;
  readonly contractId: typeof R1_WORKSTEAL_RECEIPT_SUBMISSION_CONTRACT_ID;
  readonly theoremId: typeof R1_WORKSTEAL_RECEIPT_SUBMISSION_THEOREM_ID;
  readonly executable: false;
}

function asRecord(value: unknown): Record<string, unknown> | null {
  if (!value || typeof value !== 'object' || Array.isArray(value)) {
    return null;
  }
  return value as Record<string, unknown>;
}

function isFiniteInteger(value: unknown): value is number {
  return (
    typeof value === 'number' &&
    Number.isFinite(value) &&
    Number.isInteger(value)
  );
}

function normalizeNonemptyString(value: unknown, name: string): string {
  if (typeof value !== 'string') {
    throw new Error(`${name} must be a string`);
  }
  const trimmed = value.trim();
  if (!trimmed) {
    throw new Error(`${name} must be nonempty`);
  }
  return trimmed;
}

function normalizeOptionalString(
  value: unknown,
  name: string
): string | undefined {
  if (value === undefined || value === null) {
    return undefined;
  }
  return normalizeNonemptyString(value, name);
}

function normalizeNonnegativeInteger(value: unknown, name: string): number {
  if (!isFiniteInteger(value) || value < 0) {
    throw new Error(`${name} must be a nonnegative integer`);
  }
  return value;
}

function normalizeBoundedInteger(
  value: unknown,
  name: string,
  min: number,
  max: number
): number {
  if (!isFiniteInteger(value) || value < min || value > max) {
    throw new Error(`${name} must be an integer in [${min}, ${max}]`);
  }
  return value;
}

function normalizeOptionalPriority(value: unknown): number | undefined {
  if (value === undefined || value === null) {
    return undefined;
  }
  return normalizeBoundedInteger(value, 'priority', 0, 255);
}

function normalizeWorkKind(value: unknown): 'fibonacci' | 'qwen-knot' {
  if (value !== 'fibonacci' && value !== 'qwen-knot') {
    throw new Error('workKind must be fibonacci or qwen-knot');
  }
  return value;
}

function normalizeProjection(
  projection: Protocol69ProjectionEnvelope | undefined
): Protocol69ProjectionEnvelope {
  return projection
    ? validateProtocol69ProjectionEnvelope(projection)
    : buildProtocol69ProjectionEnvelope();
}

function normalizePhyleQueueSummary(
  value: R1WorkStealPhyleQueueSummary
): R1WorkStealPhyleQueueSummary {
  return {
    phyleId: normalizeNonemptyString(value.phyleId, 'phyleId'),
    pending: normalizeNonnegativeInteger(value.pending, 'pending'),
    claimed: normalizeNonnegativeInteger(value.claimed, 'claimed'),
    accepted: normalizeNonnegativeInteger(value.accepted, 'accepted'),
    rejected: normalizeNonnegativeInteger(value.rejected, 'rejected'),
    completedUnits: normalizeNonnegativeInteger(
      value.completedUnits,
      'completedUnits'
    ),
  };
}

/**
 * Builds the R1 Work Steal Hint.
 */
export function buildR1WorkStealHint(
  input: R1WorkStealHintInput
): R1WorkStealHint {
  const priority = normalizeOptionalPriority(input.priority);
  const witnessWindowHash = normalizeOptionalString(
    input.witnessWindowHash,
    'witnessWindowHash'
  );

  const modelId = normalizeOptionalString(input.modelId, 'modelId');
  const knotId = normalizeOptionalString(input.knotId, 'knotId');
  const knotHash = normalizeOptionalString(input.knotHash, 'knotHash');
  const tokenBudget =
    input.tokenBudget === undefined
      ? undefined
      : normalizeBoundedInteger(input.tokenBudget, 'tokenBudget', 1, 4096);
  if (input.workKind === 'qwen-knot') {
    if (!modelId || !knotId || !knotHash || tokenBudget === undefined) {
      throw new Error(
        'qwen-knot hints require modelId, knotId, knotHash, and tokenBudget'
      );
    }
  }

  return {
    schemaVersion: R1_WORKSTEAL_HINT_SCHEMA_VERSION,
    contractId: R1_WORKSTEAL_HINT_CONTRACT_ID,
    theoremId: R1_WORKSTEAL_HINT_THEOREM_ID,
    executable: false,
    hintId: normalizeNonemptyString(input.hintId, 'hintId'),
    producerDeviceId: normalizeNonemptyString(
      input.producerDeviceId,
      'producerDeviceId'
    ),
    targetPhyleId: normalizeNonemptyString(input.targetPhyleId, 'targetPhyleId'),
    workKind: normalizeWorkKind(input.workKind),
    n: normalizeBoundedInteger(input.n, 'n', 0, 93),
    ...(modelId === undefined ? {} : { modelId }),
    ...(knotId === undefined ? {} : { knotId }),
    ...(knotHash === undefined ? {} : { knotHash }),
    ...(tokenBudget === undefined ? {} : { tokenBudget }),
    ...(priority === undefined ? {} : { priority }),
    ttlMs: normalizeBoundedInteger(input.ttlMs, 'ttlMs', 1, 0xffff_ffff),
    timestampMs: normalizeNonnegativeInteger(input.timestampMs, 'timestampMs'),
    ...(witnessWindowHash === undefined ? {} : { witnessWindowHash }),
    projection: normalizeProjection(input.projection),
  };
}

/**
 * Handles the gnosis validate R1 Work Steal Hint workflow.
 */
export function validateR1WorkStealHint(value: unknown): R1WorkStealHint {
  const record = asRecord(value);
  if (!record) {
    throw new Error('R1 work-steal hint must be an object');
  }
  if (record['schemaVersion'] !== R1_WORKSTEAL_HINT_SCHEMA_VERSION) {
    throw new Error('schemaVersion mismatch');
  }
  if (record['contractId'] !== R1_WORKSTEAL_HINT_CONTRACT_ID) {
    throw new Error('contractId mismatch');
  }
  if (record['theoremId'] !== R1_WORKSTEAL_HINT_THEOREM_ID) {
    throw new Error('theoremId mismatch');
  }
  if (record['executable'] !== false) {
    throw new Error('work-steal hints are non-executable');
  }

  const priority = normalizeOptionalPriority(record['priority']);
  const witnessWindowHash = normalizeOptionalString(
    record['witnessWindowHash'],
    'witnessWindowHash'
  );
  const modelId = normalizeOptionalString(record['modelId'], 'modelId');
  const knotId = normalizeOptionalString(record['knotId'], 'knotId');
  const knotHash = normalizeOptionalString(record['knotHash'], 'knotHash');
  const tokenBudget =
    record['tokenBudget'] === undefined || record['tokenBudget'] === null
      ? undefined
      : normalizeBoundedInteger(record['tokenBudget'], 'tokenBudget', 1, 4096);
  const workKind = normalizeWorkKind(record['workKind']);

  return buildR1WorkStealHint({
    hintId: normalizeNonemptyString(record['hintId'], 'hintId'),
    producerDeviceId: normalizeNonemptyString(
      record['producerDeviceId'],
      'producerDeviceId'
    ),
    targetPhyleId: normalizeNonemptyString(
      record['targetPhyleId'],
      'targetPhyleId'
    ),
    workKind,
    n: normalizeBoundedInteger(record['n'], 'n', 0, 93),
    ...(modelId === undefined ? {} : { modelId }),
    ...(knotId === undefined ? {} : { knotId }),
    ...(knotHash === undefined ? {} : { knotHash }),
    ...(tokenBudget === undefined ? {} : { tokenBudget }),
    ...(priority === undefined ? {} : { priority }),
    ttlMs: normalizeBoundedInteger(record['ttlMs'], 'ttlMs', 1, 0xffff_ffff),
    timestampMs: normalizeNonnegativeInteger(
      record['timestampMs'],
      'timestampMs'
    ),
    ...(witnessWindowHash === undefined ? {} : { witnessWindowHash }),
    projection: validateProtocol69ProjectionEnvelope(record['projection']),
  });
}

/**
 * Builds the R1 Compute Receipt.
 */
export function buildR1ComputeReceipt(
  input: R1ComputeReceiptInput
): R1ComputeReceipt {
  const witnessWindowHash = normalizeOptionalString(
    input.witnessWindowHash,
    'witnessWindowHash'
  );
  const modelId = normalizeOptionalString(input.modelId, 'modelId');
  const knotId = normalizeOptionalString(input.knotId, 'knotId');
  const knotHash = normalizeOptionalString(input.knotHash, 'knotHash');
  const tokenBudget =
    input.tokenBudget === undefined
      ? undefined
      : normalizeBoundedInteger(input.tokenBudget, 'tokenBudget', 1, 4096);

  return {
    schemaVersion: R1_COMPUTE_RECEIPT_SCHEMA_VERSION,
    contractId: R1_COMPUTE_RECEIPT_CONTRACT_ID,
    theoremId: R1_COMPUTE_RECEIPT_THEOREM_ID,
    authority: 'reuse-right',
    executable: false,
    hintId: normalizeNonemptyString(input.hintId, 'hintId'),
    workerDeviceId: normalizeNonemptyString(
      input.workerDeviceId,
      'workerDeviceId'
    ),
    targetPhyleId: normalizeNonemptyString(input.targetPhyleId, 'targetPhyleId'),
    workKind: normalizeWorkKind(input.workKind),
    n: normalizeBoundedInteger(input.n, 'n', 0, 93),
    ...(modelId === undefined ? {} : { modelId }),
    ...(knotId === undefined ? {} : { knotId }),
    ...(knotHash === undefined ? {} : { knotHash }),
    ...(tokenBudget === undefined ? {} : { tokenBudget }),
    result: normalizeNonemptyString(input.result, 'result'),
    resultDigest: normalizeNonemptyString(input.resultDigest, 'resultDigest'),
    computeNs: normalizeNonnegativeInteger(input.computeNs, 'computeNs'),
    validationNs: normalizeNonnegativeInteger(
      input.validationNs,
      'validationNs'
    ),
    timestampMs: normalizeNonnegativeInteger(input.timestampMs, 'timestampMs'),
    beamCacheKey: normalizeNonemptyString(input.beamCacheKey, 'beamCacheKey'),
    ...(witnessWindowHash === undefined ? {} : { witnessWindowHash }),
    projection: normalizeProjection(input.projection),
  };
}

/**
 * Handles the gnosis validate R1 Compute Receipt workflow.
 */
export function validateR1ComputeReceipt(value: unknown): R1ComputeReceipt {
  const record = asRecord(value);
  if (!record) {
    throw new Error('R1 compute receipt must be an object');
  }
  if (record['schemaVersion'] !== R1_COMPUTE_RECEIPT_SCHEMA_VERSION) {
    throw new Error('schemaVersion mismatch');
  }
  if (record['contractId'] !== R1_COMPUTE_RECEIPT_CONTRACT_ID) {
    throw new Error('contractId mismatch');
  }
  if (record['theoremId'] !== R1_COMPUTE_RECEIPT_THEOREM_ID) {
    throw new Error('theoremId mismatch');
  }
  if (record['authority'] !== 'reuse-right') {
    throw new Error('receipt authority must be reuse-right');
  }
  if (record['executable'] !== false) {
    throw new Error('compute receipts are non-executable');
  }

  const witnessWindowHash = normalizeOptionalString(
    record['witnessWindowHash'],
    'witnessWindowHash'
  );
  const modelId = normalizeOptionalString(record['modelId'], 'modelId');
  const knotId = normalizeOptionalString(record['knotId'], 'knotId');
  const knotHash = normalizeOptionalString(record['knotHash'], 'knotHash');
  const tokenBudget =
    record['tokenBudget'] === undefined || record['tokenBudget'] === null
      ? undefined
      : normalizeBoundedInteger(record['tokenBudget'], 'tokenBudget', 1, 4096);

  return buildR1ComputeReceipt({
    hintId: normalizeNonemptyString(record['hintId'], 'hintId'),
    workerDeviceId: normalizeNonemptyString(
      record['workerDeviceId'],
      'workerDeviceId'
    ),
    targetPhyleId: normalizeNonemptyString(
      record['targetPhyleId'],
      'targetPhyleId'
    ),
    workKind: normalizeWorkKind(record['workKind']),
    n: normalizeBoundedInteger(record['n'], 'n', 0, 93),
    ...(modelId === undefined ? {} : { modelId }),
    ...(knotId === undefined ? {} : { knotId }),
    ...(knotHash === undefined ? {} : { knotHash }),
    ...(tokenBudget === undefined ? {} : { tokenBudget }),
    result: normalizeNonemptyString(record['result'], 'result'),
    resultDigest: normalizeNonemptyString(record['resultDigest'], 'resultDigest'),
    computeNs: normalizeNonnegativeInteger(record['computeNs'], 'computeNs'),
    validationNs: normalizeNonnegativeInteger(
      record['validationNs'],
      'validationNs'
    ),
    timestampMs: normalizeNonnegativeInteger(
      record['timestampMs'],
      'timestampMs'
    ),
    beamCacheKey: normalizeNonemptyString(record['beamCacheKey'], 'beamCacheKey'),
    ...(witnessWindowHash === undefined ? {} : { witnessWindowHash }),
    projection: validateProtocol69ProjectionEnvelope(record['projection']),
  });
}

/**
 * Builds the R1 Work Steal Queue Summary.
 */
export function buildR1WorkStealQueueSummary(
  input: R1WorkStealQueueSummaryInput
): R1WorkStealQueueSummary {
  return {
    schemaVersion: R1_WORKSTEAL_QUEUE_SUMMARY_SCHEMA_VERSION,
    contractId: R1_WORKSTEAL_QUEUE_SUMMARY_CONTRACT_ID,
    pending: normalizeNonnegativeInteger(input.pending, 'pending'),
    claimed: normalizeNonnegativeInteger(input.claimed, 'claimed'),
    accepted: normalizeNonnegativeInteger(input.accepted, 'accepted'),
    rejected: normalizeNonnegativeInteger(input.rejected, 'rejected'),
    perPhyle: input.perPhyle.map(normalizePhyleQueueSummary),
    timestampMs: normalizeNonnegativeInteger(input.timestampMs, 'timestampMs'),
  };
}

/**
 * Handles the gnosis validate R1 Work Steal Queue Summary workflow.
 */
export function validateR1WorkStealQueueSummary(
  value: unknown
): R1WorkStealQueueSummary {
  const record = asRecord(value);
  if (!record) {
    throw new Error('R1 work-steal queue summary must be an object');
  }
  if (record['schemaVersion'] !== R1_WORKSTEAL_QUEUE_SUMMARY_SCHEMA_VERSION) {
    throw new Error('schemaVersion mismatch');
  }
  if (record['contractId'] !== R1_WORKSTEAL_QUEUE_SUMMARY_CONTRACT_ID) {
    throw new Error('contractId mismatch');
  }
  const perPhyle = record['perPhyle'];
  if (!Array.isArray(perPhyle)) {
    throw new Error('perPhyle must be an array');
  }
  return buildR1WorkStealQueueSummary({
    pending: normalizeNonnegativeInteger(record['pending'], 'pending'),
    claimed: normalizeNonnegativeInteger(record['claimed'], 'claimed'),
    accepted: normalizeNonnegativeInteger(record['accepted'], 'accepted'),
    rejected: normalizeNonnegativeInteger(record['rejected'], 'rejected'),
    perPhyle: perPhyle.map((entry) => {
      const phyleRecord = asRecord(entry);
      if (!phyleRecord) {
        throw new Error('perPhyle entries must be objects');
      }
      return {
        phyleId: normalizeNonemptyString(phyleRecord['phyleId'], 'phyleId'),
        pending: normalizeNonnegativeInteger(phyleRecord['pending'], 'pending'),
        claimed: normalizeNonnegativeInteger(phyleRecord['claimed'], 'claimed'),
        accepted: normalizeNonnegativeInteger(
          phyleRecord['accepted'],
          'accepted'
        ),
        rejected: normalizeNonnegativeInteger(
          phyleRecord['rejected'],
          'rejected'
        ),
        completedUnits: normalizeNonnegativeInteger(
          phyleRecord['completedUnits'],
          'completedUnits'
        ),
      };
    }),
    timestampMs: normalizeNonnegativeInteger(
      record['timestampMs'],
      'timestampMs'
    ),
  });
}

/**
 * Builds the R1 Work Steal Claim.
 */
export function buildR1WorkStealClaim(
  input: R1WorkStealClaimInput
): R1WorkStealClaim {
  const claimId = normalizeOptionalString(input.claimId, 'claimId');
  const hint =
    input.hint === undefined || input.hint === null
      ? undefined
      : validateR1WorkStealHint(input.hint);
  if (input.status === 'claimed' && (!claimId || !hint)) {
    throw new Error('claimed work-steal responses require claimId and hint');
  }
  if (input.status === 'idle' && (claimId || hint)) {
    throw new Error('idle work-steal responses must not include claimId or hint');
  }
  return {
    schemaVersion: R1_WORKSTEAL_CLAIM_SCHEMA_VERSION,
    contractId: R1_WORKSTEAL_CLAIM_CONTRACT_ID,
    theoremId: R1_WORKSTEAL_CLAIM_THEOREM_ID,
    executable: false,
    status: input.status,
    workerDeviceId: normalizeNonemptyString(
      input.workerDeviceId,
      'workerDeviceId'
    ),
    phyleId: normalizeNonemptyString(input.phyleId, 'phyleId'),
    ...(claimId === undefined ? {} : { claimId }),
    ...(hint === undefined ? {} : { hint }),
    summary: validateR1WorkStealQueueSummary(input.summary),
    timestampMs: normalizeNonnegativeInteger(input.timestampMs, 'timestampMs'),
    projection: normalizeProjection(input.projection),
  };
}

/**
 * Handles the gnosis validate R1 Work Steal Claim workflow.
 */
export function validateR1WorkStealClaim(value: unknown): R1WorkStealClaim {
  const record = asRecord(value);
  if (!record) {
    throw new Error('R1 work-steal claim must be an object');
  }
  if (record['schemaVersion'] !== R1_WORKSTEAL_CLAIM_SCHEMA_VERSION) {
    throw new Error('schemaVersion mismatch');
  }
  if (record['contractId'] !== R1_WORKSTEAL_CLAIM_CONTRACT_ID) {
    throw new Error('contractId mismatch');
  }
  if (record['theoremId'] !== R1_WORKSTEAL_CLAIM_THEOREM_ID) {
    throw new Error('theoremId mismatch');
  }
  if (record['executable'] !== false) {
    throw new Error('work-steal claims are non-executable');
  }
  if (record['status'] !== 'claimed' && record['status'] !== 'idle') {
    throw new Error('status must be claimed or idle');
  }
  return buildR1WorkStealClaim({
    status: record['status'],
    workerDeviceId: normalizeNonemptyString(
      record['workerDeviceId'],
      'workerDeviceId'
    ),
    phyleId: normalizeNonemptyString(record['phyleId'], 'phyleId'),
    ...(record['claimId'] === undefined || record['claimId'] === null
      ? {}
      : { claimId: normalizeNonemptyString(record['claimId'], 'claimId') }),
    ...(record['hint'] === undefined || record['hint'] === null
      ? {}
      : { hint: validateR1WorkStealHint(record['hint']) }),
    summary: validateR1WorkStealQueueSummary(record['summary']),
    timestampMs: normalizeNonnegativeInteger(
      record['timestampMs'],
      'timestampMs'
    ),
    projection: validateProtocol69ProjectionEnvelope(record['projection']),
  });
}

/**
 * Builds the R1 Work Steal Receipt Submission.
 */
export function buildR1WorkStealReceiptSubmission(
  input: R1WorkStealReceiptSubmissionInput
): R1WorkStealReceiptSubmission {
  return {
    schemaVersion: R1_WORKSTEAL_RECEIPT_SUBMISSION_SCHEMA_VERSION,
    contractId: R1_WORKSTEAL_RECEIPT_SUBMISSION_CONTRACT_ID,
    theoremId: R1_WORKSTEAL_RECEIPT_SUBMISSION_THEOREM_ID,
    executable: false,
    claimId: normalizeNonemptyString(input.claimId, 'claimId'),
    receipt: validateR1ComputeReceipt(input.receipt),
    timestampMs: normalizeNonnegativeInteger(input.timestampMs, 'timestampMs'),
    projection: normalizeProjection(input.projection),
  };
}

/**
 * Handles the gnosis validate R1 Work Steal Receipt Submission workflow.
 */
export function validateR1WorkStealReceiptSubmission(
  value: unknown
): R1WorkStealReceiptSubmission {
  const record = asRecord(value);
  if (!record) {
    throw new Error('R1 work-steal receipt submission must be an object');
  }
  if (
    record['schemaVersion'] !== R1_WORKSTEAL_RECEIPT_SUBMISSION_SCHEMA_VERSION
  ) {
    throw new Error('schemaVersion mismatch');
  }
  if (record['contractId'] !== R1_WORKSTEAL_RECEIPT_SUBMISSION_CONTRACT_ID) {
    throw new Error('contractId mismatch');
  }
  if (record['theoremId'] !== R1_WORKSTEAL_RECEIPT_SUBMISSION_THEOREM_ID) {
    throw new Error('theoremId mismatch');
  }
  if (record['executable'] !== false) {
    throw new Error('receipt submissions are non-executable');
  }
  return buildR1WorkStealReceiptSubmission({
    claimId: normalizeNonemptyString(record['claimId'], 'claimId'),
    receipt: validateR1ComputeReceipt(record['receipt']),
    timestampMs: normalizeNonnegativeInteger(
      record['timestampMs'],
      'timestampMs'
    ),
    projection: validateProtocol69ProjectionEnvelope(record['projection']),
  });
}
