// @a0n/.ignored_gnosis/src/quantum-handlers.ts
/**
 * Quantum GG Topology Handlers
 *
 * Handlers for quantum circuit construction, gate application,
 * measurement, and entanglement detection in GG topologies.
 */

import type { GnosisRegistry } from './runtime/registry';
import {
  ket0,
  ket1,
  ketPlus,
  ketMinus,
  hadamard,
  pauliX,
  pauliZ,
  measure,
  isNormalized,
  prob0,
  prob1,
  createRegister,
  registerFromQubit,
  applyCNOT,
  applyCZ,
  applyGateToRegister,
  measureRegister,
  bellState,
  isEntangled,
  registerNormalized,
  createCircuit,
  addGate,
  addCNOT,
  addMeasure,
  addBarrier,
  executeCircuit,
  H_GATE,
  X_GATE,
  Z_GATE,
  S_GATE,
  T_GATE,
  phaseGate,
  type Qubit,
  type QuantumRegister,
} from './quantum';

let rngSeed = 42;
function defaultRng(): number {
  rngSeed = (rngSeed * 1103515245 + 12345) & 0x7fffffff;
  return rngSeed / 0x7fffffff;
}

/**
 * Handles the gnosis register Quantum Handlers workflow.
 */
export function registerQuantumHandlers(registry: GnosisRegistry): void {
  // Initialize qubit in a basis state
  registry.register('QubitInit', async (_payload, props) => {
    const state = props.state ?? '0';
    switch (state) {
      case '0':
        return ket0();
      case '1':
        return ket1();
      case '+':
        return ketPlus();
      case '-':
        return ketMinus();
      default:
        return ket0();
    }
  });

  // Apply a single-qubit gate
  registry.register('QuantumGate', async (payload, props) => {
    const qubit = payload as Qubit;
    const gate = props.gate ?? 'H';
    switch (gate) {
      case 'H':
        return hadamard(qubit);
      case 'X':
        return pauliX(qubit);
      case 'Z':
        return pauliZ(qubit);
      default:
        return hadamard(qubit);
    }
  });

  // Measure a qubit (OBSERVE semantics)
  registry.register('QuantumMeasure', async (payload) => {
    const qubit = payload as Qubit;
    return measure(qubit, defaultRng);
  });

  // Create a quantum register
  registry.register('RegisterInit', async (_payload, props) => {
    const numQubits = parseInt(props.qubits ?? '2', 10);
    return createRegister(numQubits);
  });

  // Apply gate to register qubit
  registry.register('RegisterGate', async (payload, props) => {
    const reg = payload as QuantumRegister;
    const qubitIdx = parseInt(props.qubit ?? '0', 10);
    const gate = props.gate ?? 'H';
    const gateMatrix =
      gate === 'X'
        ? X_GATE
        : gate === 'Z'
        ? Z_GATE
        : gate === 'S'
        ? S_GATE
        : gate === 'T'
        ? T_GATE
        : H_GATE;
    return applyGateToRegister(reg, qubitIdx, gateMatrix);
  });

  // CNOT gate on register
  registry.register('CNOT', async (payload, props) => {
    const reg = payload as QuantumRegister;
    const control = parseInt(props.control ?? '0', 10);
    const target = parseInt(props.target ?? '1', 10);
    return applyCNOT(reg, control, target);
  });

  // CZ gate on register
  registry.register('CZ', async (payload, props) => {
    const reg = payload as QuantumRegister;
    const control = parseInt(props.control ?? '0', 10);
    const target = parseInt(props.target ?? '1', 10);
    return applyCZ(reg, control, target);
  });

  // Measure register qubit
  registry.register('RegisterMeasure', async (payload, props) => {
    const reg = payload as QuantumRegister;
    const qubitIdx = parseInt(props.qubit ?? '0', 10);
    return measureRegister(reg, qubitIdx, defaultRng);
  });

  // Create Bell state (entangled pair)
  registry.register('BellState', async () => {
    return bellState();
  });

  // Check entanglement
  registry.register('EntanglementCheck', async (payload) => {
    const reg = payload as QuantumRegister;
    return {
      entangled: isEntangled(reg),
      normalized: registerNormalized(reg),
      numQubits: reg.numQubits,
    };
  });

  // Build and execute a circuit
  registry.register('QuantumCircuitExecutor', async (payload, props) => {
    const numQubits = parseInt(props.qubits ?? '2', 10);
    const opsRaw = props.ops ?? 'H0,CNOT01,M0,M1';
    const circuit = createCircuit(numQubits);

    for (const op of opsRaw.split(',')) {
      const trimmed = op.trim();
      if (trimmed.startsWith('H')) {
        addGate(circuit, parseInt(trimmed.slice(1), 10), H_GATE, 'H');
      } else if (trimmed.startsWith('X')) {
        addGate(circuit, parseInt(trimmed.slice(1), 10), X_GATE, 'X');
      } else if (trimmed.startsWith('Z')) {
        addGate(circuit, parseInt(trimmed.slice(1), 10), Z_GATE, 'Z');
      } else if (trimmed.startsWith('CNOT')) {
        const c = parseInt(trimmed[4], 10);
        const t = parseInt(trimmed[5], 10);
        addCNOT(circuit, c, t);
      } else if (trimmed.startsWith('M')) {
        addMeasure(circuit, parseInt(trimmed.slice(1), 10));
      } else if (trimmed === 'B') {
        addBarrier(circuit);
      }
    }

    const result = executeCircuit(circuit, defaultRng);
    return {
      measurements: Object.fromEntries(result.measurements),
      numQubits,
      opsCount: circuit.ops.length,
      normalized: registerNormalized(result.register),
    };
  });
}

export const QUANTUM_GATE_NAMES = [
  'H',
  'X',
  'Y',
  'Z',
  'S',
  'T',
  'CNOT',
  'CZ',
] as const;
