// @a0n/.ignored_gnosis/src/quasicrystalline-clock.ts
/**
 * quasicrystalline-clock.ts -- Pisot-native distributed synchronization.
 *
 * Canonical spec (source of truth):
 *   open-source/gnosis/lean/Lean/ForkRaceFoldTheorems/QuasicrystallineClock.lean
 *   open-source/gnosis/lean/Lean/ForkRaceFoldTheorems/PisotVickreyTable.lean
 *
 * Canonical WASM/TS kernel for FORK/RACE/CHECK:
 *   open-source/aether/src/wasm-simd/pisot-cpu.ts
 *
 * Indexing convention: Lean Fibonacci with F[0] = F[1] = 1, F[n+2] = F[n+1] + F[n].
 * The Fibonacci backing is delegated to `pisotF` from ./betti/pisot-primitive.ts,
 * which mirrors PisotVickreyTable.F byte-for-byte so fork/race/Cassini identities
 * (frf_master, cassini_no_silent_corruption) hold without re-derivation.
 *
 * A distributed clock built on a 3D Pisot lattice (phi x Pell x Tribonacci)
 * where synchronization is not a single linear "tick", but a vector snapping
 * mechanism that tolerates structural jitter via the Cassini gap.
 */

import { pisotF } from './betti/pisot-primitive';
import { ClinamenTier, getClinamenCost } from './gnostic-time';
// ============================================================================
// Pisot Towers (φ, Pell, Tribonacci)
// ============================================================================

/** Cap indices to keep values within reasonable BigInt bounds. */
export const CLOCK_MAX_INDEX = 64;

/** Pell Tower: P[0]=1, P[1]=2, P[n+2] = 2P[n+1] + P[n]. Ratio ≈ 2.414. */
export const PELL_TOWER: ReadonlyArray<bigint> = (() => {
  const p = new Array<bigint>(CLOCK_MAX_INDEX + 1);
  p[0] = 1n;
  p[1] = 2n;
  for (let n = 2; n <= CLOCK_MAX_INDEX; n++) {
    const prev = p[n - 1];
    const prevPrev = p[n - 2];
    if (prev === undefined || prevPrev === undefined) {
      throw new Error(`[Clock] Pell seed missing at index ${n}`);
    }
    p[n] = 2n * prev + prevPrev;
  }
  return p;
})();

/** Tribonacci Tower: T[0]=1, T[1]=1, T[2]=2, T[n+3] = T[n+2] + T[n+1] + T[n]. Ratio ≈ 1.839. */
export const TRIB_TOWER: ReadonlyArray<bigint> = (() => {
  const t = new Array<bigint>(CLOCK_MAX_INDEX + 1);
  t[0] = 1n;
  t[1] = 1n;
  t[2] = 2n;
  for (let n = 3; n <= CLOCK_MAX_INDEX; n++) {
    const prev = t[n - 1];
    const prevPrev = t[n - 2];
    const prevThird = t[n - 3];
    if (prev === undefined || prevPrev === undefined || prevThird === undefined) {
      throw new Error(`[Clock] Tribonacci seed missing at index ${n}`);
    }
    t[n] = prev + prevPrev + prevThird;
  }
  return t;
})();

function requireTowerValue(
  tower: ReadonlyArray<bigint>,
  index: number,
  label: string
): bigint {
  const value = tower[index];
  if (value === undefined) {
    throw new Error(`[Clock] ${label} tower missing index ${index}`);
  }
  return value;
}

// ============================================================================
// 3D Pisot Vector Clock
// ============================================================================

export interface PisotClock {
  phi: number;
  pell: number;
  trib: number;
}

/**
 * The "Vickrey Moment" (Clock Projection) — the scalar shadow
 * cast by the 3D coordinate onto the natural numbers.
 */
export function vickreyMoment(clock: PisotClock): bigint {
  const f = BigInt(pisotF(clock.phi));
  const p = requireTowerValue(PELL_TOWER, clock.pell, 'pell');
  const t = requireTowerValue(TRIB_TOWER, clock.trib, 'trib');
  return f * p * t;
}

// ============================================================================
// Cassini Parity Filter (Phase Alternation)
// ============================================================================

export interface ClockPulse {
  prev: bigint;
  curr: bigint;
  next: bigint;
}

export enum PulsePhase {
  Even = 'even',
  Odd = 'odd',
  Invalid = 'invalid',
}

/**
 * Evaluates the Cassini phase of a Fibonacci triple.
 * Mirrors `pulsePhase` in QuasicrystallineClock.lean.
 */
export function evaluatePhase(p: ClockPulse): PulsePhase {
  if (p.next * p.prev === p.curr * p.curr + 1n) return PulsePhase.Even;
  if (p.curr * p.curr === p.next * p.prev + 1n) return PulsePhase.Odd;
  return PulsePhase.Invalid;
}

/** A 3D pulse captures the state triple for all three gears. */
export interface Pulse3D {
  phi: ClockPulse;
  pell: ClockPulse;
  trib: ClockPulse;
}

/**
 * Kernel-level gatekeeper (isValidPulse3D in Lean).
 * Returns true if the Phi and Pell components satisfy the gap-1 Cassini phase structure.
 * Note: Tribonacci requires a deeper determinant check and is bypassed in this v1.
 */
export function isValidPulse3D(p: Pulse3D): boolean {
  return (
    evaluatePhase(p.phi) !== PulsePhase.Invalid &&
    evaluatePhase(p.pell) !== PulsePhase.Invalid
    // trib is currently a pass-through until 3-step Cassini is implemented
  );
}

/**
 * The Topological Snapping Mechanism (FOLD).
 * If a value is within the Cassini Margin (+/- 1) of a target lattice coordinate,
 * we "snap" it back to that exact coordinate.
 * 
 * The `tier` parameter realizes the Resolution Choice Theorem: You choose the 
 * temporal resolution of the Clinamen tax, but the topological cost is ALWAYS 1.
 */
export function snapToLattice(
  received: bigint,
  expected: bigint,
  _tier: ClinamenTier = ClinamenTier.Second
): bigint {
  const clinamenCost = getClinamenCost(_tier);
  if (received === expected + clinamenCost || received === expected - clinamenCost) {
    return expected;
  }
  return received;
}

/**
 * Validates whether two clocks are within the Cassini Vickrey Margin.
 */
export function isSynchronized(
  val: bigint,
  expected: bigint,
  _tier: ClinamenTier = ClinamenTier.Second
): boolean {
  const clinamenCost = getClinamenCost(_tier);
  return (
    val === expected ||
    val === expected + clinamenCost ||
    expected === val + clinamenCost
  );
}

/**
 * Teleportation jump (FORK) primitive.
 * Instead of linear ticks, jump state by applying Fibonacci product.
 */
export function pisotTeleport(localIndex: number, gapIndex: number): number {
  // In the real system, this would return the index F(m+k+2).
  // For the clock implementation, we treat it as an index shift.
  return localIndex + gapIndex + 2;
}

/**
 * Verification of a future state via the FORK identity.
 * Mirrors `zeroTrustSync` in QuasicrystallineClock.lean.
 */
export function verifyTeleport(
  localM: number,
  gapK: number,
  receivedNext: bigint
): boolean {
  const fm = BigInt(pisotF(localM));
  const fm1 = BigInt(pisotF(localM + 1));
  const fk = BigInt(pisotF(gapK));
  const fk1 = BigInt(pisotF(gapK + 1));
  
  // phiFork(m, k) = F[m+1]*F[k+1] + F[m]*F[k]
  const calculatedNext = fm1 * fk1 + fm * fk;
  return calculatedNext === receivedNext;
}
