// @a0n/.ignored_gnosis/src/quasicrystalline-physical-clock.ts
/**
 * quasicrystalline-physical-clock.ts -- Operational variant of QuasicrystallineClock
 * that uses real wall-clock timestamps as the trajectory.
 *
 * Logical spec (source of truth):
 *   open-source/gnosis/lean/Lean/ForkRaceFoldTheorems/QuasicrystallineClock.lean
 *   In particular drift_bounded_k_rounds (§8): for k rounds with per-step
 *   Lipschitz bound 1, the cumulative drift between two honest clocks is ≤ k.
 *
 * Encoding:
 *   wall-clock nanoseconds  →  lattice index (bucket = LATTICE_SCALE_NS)
 *   index n                 →  Cassini witness triple ( F[n-1], F[n], F[n+1] )
 *
 * The lattice scale is configured so that one lattice unit ≈ 100ms, which
 * dominates typical NTP jitter (±50ms) -- so two honest clocks polled at
 * the same physical instant land within ±1 lattice unit (the Cassini margin).
 *
 * This module is additive: it does not modify quasicrystalline-clock.ts.
 */

import { pisotF, pisotCassini } from './betti/pisot-primitive';
import { evaluatePhase, PulsePhase, type ClockPulse } from './quasicrystalline-clock';

// ============================================================================
// Configuration
// ============================================================================

/** One lattice unit ≈ 100ms = 100_000_000 ns. NTP jitter (±50ms) sits well inside. */
export const LATTICE_SCALE_NS: bigint = 100_000_000n;

/** Default max acceptable drift, in lattice units (≈100s at 100ms/unit). */
export const DEFAULT_MAX_DRIFT = 1000;

/** Cap to avoid overflowing the Fibonacci table in pisot-primitive (~63). */
export const PHYSICAL_CLOCK_INDEX_CAP = 60;

// ============================================================================
// Types
// ============================================================================

export interface PhysicalClockReading {
  /** Raw nanosecond timestamp from process.hrtime.bigint() (relative to clock epoch). */
  timestamp: bigint;
  /** Pisot lattice index n = floor((timestamp - epoch) / LATTICE_SCALE_NS) mod cap. */
  pisotIndex: number;
  /** Cassini witness triple at index n: ( F[n-1], F[n], F[n+1] ). */
  cassiniWitness: ClockPulse;
}

export type ValidationSource = 'cassini' | 'drift' | null;

export interface ValidationResult {
  valid: boolean;
  drift: number;
  source: ValidationSource;
}

// ============================================================================
// Encoding helpers
// ============================================================================

/** Map a (relative) ns timestamp into a lattice index in [0, cap]. */
export function encodeIndex(deltaNs: bigint, cap = PHYSICAL_CLOCK_INDEX_CAP): number {
  if (deltaNs < 0n) deltaNs = 0n;
  const bucket = Number(deltaNs / LATTICE_SCALE_NS);
  // Wrap into the lattice cap so we never overflow Fibonacci storage.
  return ((bucket % cap) + cap) % cap;
}

/** Build the Cassini witness triple for a lattice index. */
export function witnessAt(index: number): ClockPulse {
  const n = Math.max(1, Math.min(PHYSICAL_CLOCK_INDEX_CAP - 1, index));
  return {
    prev: BigInt(pisotF(n - 1)),
    curr: BigInt(pisotF(n)),
    next: BigInt(pisotF(n + 1)),
  };
}

// ============================================================================
// PhysicalQuasicrystallineClock
// ============================================================================

export interface PhysicalClockOptions {
  maxDrift?: number;
  /** Inject a custom time source for testing. Must return ns. */
  timeSource?: () => bigint;
  /** Trajectory ring-buffer capacity. */
  trajectoryCapacity?: number;
}

export class PhysicalQuasicrystallineClock {
  readonly maxDrift: number;
  readonly epoch: bigint;
  private readonly timeSource: () => bigint;
  private readonly trajectory: PhysicalClockReading[] = [];
  private readonly trajectoryCapacity: number;

  constructor(opts: PhysicalClockOptions = {}) {
    this.maxDrift = opts.maxDrift ?? DEFAULT_MAX_DRIFT;
    this.timeSource = opts.timeSource ?? (() => process.hrtime.bigint());
    this.trajectoryCapacity = opts.trajectoryCapacity ?? 1024;
    this.epoch = this.timeSource();
  }

  /** Read the current physical clock and emit a Pisot lattice reading. */
  now(): PhysicalClockReading {
    const ts = this.timeSource();
    const delta = ts - this.epoch;
    const index = encodeIndex(delta);
    const reading: PhysicalClockReading = {
      timestamp: ts,
      pisotIndex: index,
      cassiniWitness: witnessAt(index),
    };
    this.trajectory.push(reading);
    if (this.trajectory.length > this.trajectoryCapacity) {
      this.trajectory.shift();
    }
    return reading;
  }

  /**
   * Validate a remote reading against this clock's current position.
   * Two-stage gate:
   *   1. Cassini witness must be a valid Fibonacci triple (gap-1 phase).
   *   2. Lattice-index drift must be ≤ maxDrift.
   */
  validate(remote: PhysicalClockReading): ValidationResult {
    // Stage 1: Cassini structural check on the witness triple itself.
    const phase = evaluatePhase(remote.cassiniWitness);
    if (phase === PulsePhase.Invalid) {
      return { valid: false, drift: Number.POSITIVE_INFINITY, source: 'cassini' };
    }

    // Cross-check that the witness actually corresponds to the claimed index.
    const expected = witnessAt(remote.pisotIndex);
    if (
      expected.prev !== remote.cassiniWitness.prev ||
      expected.curr !== remote.cassiniWitness.curr ||
      expected.next !== remote.cassiniWitness.next
    ) {
      return { valid: false, drift: Number.POSITIVE_INFINITY, source: 'cassini' };
    }

    // Stage 2: drift check vs the local clock right now.
    const local = this.now();
    const drift = circularDistance(local.pisotIndex, remote.pisotIndex, PHYSICAL_CLOCK_INDEX_CAP);
    if (drift > this.maxDrift) {
      return { valid: false, drift, source: 'drift' };
    }

    return { valid: true, drift, source: null };
  }

  /**
   * Sync to a remote reading. If the remote validates and is ahead of us by
   * more than 0 lattice units, we shift our epoch backward to align.
   * Returns the validation result that was used.
   */
  sync(remote: PhysicalClockReading): ValidationResult {
    const result = this.validate(remote);
    if (!result.valid) return result;
    const local = this.now();
    if (remote.pisotIndex > local.pisotIndex) {
      // Pull the epoch back so our next now() lands on the remote index.
      const shift = BigInt(remote.pisotIndex - local.pisotIndex) * LATTICE_SCALE_NS;
      (this as { epoch: bigint }).epoch = this.epoch - shift;
    }
    return result;
  }

  /** Last `window` readings. */
  getTrajectory(window: number): PhysicalClockReading[] {
    if (window <= 0) return [];
    return this.trajectory.slice(-window);
  }
}

// ============================================================================
// Drift utilities
// ============================================================================

/**
 * Circular distance on a Z/cap lattice. Mirrors the Lipschitz-1 metric used
 * in drift_bounded_k_rounds: each round shifts the index by at most 1.
 */
export function circularDistance(a: number, b: number, cap = PHYSICAL_CLOCK_INDEX_CAP): number {
  const raw = Math.abs(a - b);
  return Math.min(raw, cap - raw);
}

/**
 * Total path-length of a trajectory under the Lipschitz-1 metric.
 * The Lean theorem `drift_bounded_k_rounds` states this is ≤ k for k rounds.
 */
export function trajectoryDrift(readings: PhysicalClockReading[]): number {
  let total = 0;
  for (let i = 1; i < readings.length; i++) {
    total += circularDistance(readings[i - 1].pisotIndex, readings[i].pisotIndex);
  }
  return total;
}

/** Re-export Cassini primitive for callers that want to verify by hand. */
export { pisotCassini };
