// @a0n/.ignored_gnosis/src/refinement-contagion.ts
import { QBracket, getPhiTowerStep } from './bracketed-space';
import { SpacetimeEvent, spacetimeInterval } from './causal-topology';
import { LIGHT_PICOLORENZO_LY } from './math';

/**
 * A Refinement Wave: a signal emitted when a node narrows its bracket.
 */
export interface RefinementWave {
  readonly origin: SpacetimeEvent;
  readonly bracket: QBracket;
  readonly towerStep: number;
}

/**
 * A Mesh Node that maintains a localized Bracketed Real.
 */
export class GnosticNode {
  public localBracket: QBracket;
  public localTime: number = 0;
  public buleBalance: number = 100; // Starting budget

  constructor(
    public readonly id: string,
    public readonly position: { x: number; y: number; z: number }
  ) {
    // Start with a coarse bracket
    this.localBracket = getPhiTowerStep(0);
  }

  get spacetimeEvent(): SpacetimeEvent {
    return {
      id: this.id,
      label: `Node ${this.id}`,
      ...this.position,
      pLo: this.localTime
    };
  }

  /**
   * Narrow the local bracket (consuming computational energy).
   */
  refine(n: number) {
    const result = this.localBracket.refine(n, this.buleBalance);
    this.localBracket = result.bracket;
    this.buleBalance -= result.cost;
    console.log(`[Node ${this.id}] Refined to Step ${n} (Cost: ${result.cost} Bule, Remaining: ${this.buleBalance})`);
  }

  /**
   * Emit a refinement wave to the Mesh.
   */
  emit(n: number): RefinementWave {
    this.refine(n);
    return {
      origin: this.spacetimeEvent,
      bracket: this.localBracket,
      towerStep: n
    };
  }

  /**
   * Trade precision for Bule.
   * Node B buys the narrowed bracket from Node A.
   */
  buy(seller: GnosticNode, price: number) {
    if (this.buleBalance < price) {
      console.log(`[Node ${this.id}] Cannot afford trade (Balance: ${this.buleBalance})`);
      return;
    }

    // Arbitrage check: is it cheaper than local computation?
    const localCost = 5; // simplified local cost for step 5
    if (price >= localCost) {
      console.log(`[Node ${this.id}] Trade rejected: local computation is cheaper.`);
      return;
    }

    console.log(`[Node ${this.id}] BUYING Step 5 from ${seller.id} for ${price} Bule.`);
    this.localBracket = seller.localBracket;
    this.buleBalance -= price;
    seller.buleBalance += price;
  }

  /**
   * Process an incoming wave from another node.
   * Consensus is achieved via FOLD (Intersection).
   */
  receive(wave: RefinementWave) {
    const interval = spacetimeInterval(wave.origin, this.spacetimeEvent);

    // Relativistic check: has the signal reached us?
    if (interval.temporalSeparationPLo < interval.lightCrossingPLo) {
      console.log(`[Node ${this.id}] Wave from ${wave.origin.id} has not arrived yet (Latency Gap).`);
      return;
    }

    // FOLD: Intersect our local bracket with the incoming refinement.
    // In our system, Step N+1 is always inside Step N, so we keep the narrowest.
    const incomingWidth = Number(wave.bracket.width.num) / Number(wave.bracket.width.den);
    const localWidth = Number(this.localBracket.width.num) / Number(this.localBracket.width.den);

    if (incomingWidth < localWidth) {
      console.log(`[Node ${this.id}] FOLDING with narrower bracket from ${wave.origin.id} (Sliver Width: ${incomingWidth.toFixed(10)})`);
      this.localBracket = wave.bracket;
    } else {
      console.log(`[Node ${this.id}] Local bracket is already narrower (Noise Vented).`);
    }
  }
}

/**
 * Demonstration of Relativistic Refinement Contagion.
 */
export function runContagionDemo() {
  const nodeA = new GnosticNode("A", { x: 0, y: 0, z: 0 });
  const nodeB = new GnosticNode("B", { x: 2 * LIGHT_PICOLORENZO_LY, y: 0, z: 0 }); // 2 lpLo away

  console.log("Initial state: both nodes at Step 0 refinement.\n");

  // Node A refines to Step 5 and emits a wave at T=0.
  const wave = nodeA.emit(5);
  console.log(`Node A emitted Step 5 refinement at T=0.\n`);

  // Node B tries to receive the wave at T=1.
  nodeB.localTime = 1;
  console.log(`Node B attempts to receive at T=1 (Distance: 2 lpLo):`);
  nodeB.receive(wave);

    // Node B tries to receive the wave at T=3.

    nodeB.localTime = 3;

    console.log(`\nNode B attempts to receive at T=3 (Signal has arrived):`);

    nodeB.receive(wave);

  

    console.log(`\n=== ENERGY MARKET TRADE ===`);

    const richNode = new GnosticNode("Rich", { x: 0, y: 0, z: 0 });

    const poorNode = new GnosticNode("Poor", { x: 0, y: 0, z: 0 });

    poorNode.buleBalance = 10;

  

    richNode.refine(5);

    console.log(`Rich Node Balance: ${richNode.buleBalance}, Sliver: ${richNode.localBracket.sliver.toFixed(10)}`);

    console.log(`Poor Node Balance: ${poorNode.buleBalance}, Sliver: ${poorNode.localBracket.sliver.toFixed(10)}`);

  

    poorNode.buy(richNode, 2);

  

    console.log(`\nAfter Trade:`);

    console.log(`Rich Node Balance: ${richNode.buleBalance}`);

    console.log(`Poor Node Balance: ${poorNode.buleBalance}, Sliver: ${poorNode.localBracket.sliver.toFixed(10)}`);

  }
