// @a0n/.ignored_gnosis/src/reinforcement-topology.ts
/**
 * Reinforcement Topology -- Weight Amplification on Fixed Knots
 *
 * TypeScript mirror of ReinforcementTopology.lean.
 *
 * Reinforcement pours energy into an existing topology to amplify its
 * weight without altering its crossing structure.  It is the dual of
 * Projection (creating topology from the void).
 *
 * The Hiawatha Effect governs reinforcement: energy must meander through
 * the object's internal pathways.  A straight-line dump shatters the target.
 */

// ---------------------------------------------------------------------------
// Part I: The Target Topology
// ---------------------------------------------------------------------------

export interface TargetObject {
  /** Internal topological complexity (molecular knot diagram) */
  crossings: number;
  /** Current performance level (sharpness, hardness, brightness) */
  weight: number;
  /** Maximum energy the structure can absorb before rupture */
  capacity: number;
  /** Knot invariant (identity of the object) */
  invariant: number;
}

export interface StructuralAnalysis {
  /** Crossings the mage has mapped */
  mappedCrossings: number;
  /** Actual crossings in the object */
  totalCrossings: number;
  /** Accuracy percentage (0-100) */
  accuracy: number;
}

/** A perfect analysis maps every crossing. */
export function isPerfectAnalysis(sa: StructuralAnalysis): boolean {
  return sa.mappedCrossings === sa.totalCrossings && sa.accuracy === 100;
}

/** Unmapped crossings = blind spots where energy hits dead ends. */
export function blindSpots(sa: StructuralAnalysis): number {
  return Math.max(0, sa.totalCrossings - sa.mappedCrossings);
}

// ---------------------------------------------------------------------------
// Part II: Reinforcement
// ---------------------------------------------------------------------------

export interface ReinforcementAttempt {
  target: TargetObject;
  analysis: StructuralAnalysis;
  energy: number;
}

export type ReinforcementOutcome =
  | { kind: "success"; newWeight: number }
  | { kind: "shatter" }
  | { kind: "fizzle" };

/** Reinforced weight: base + energy routed through mapped pathways. */
export function reinforcedWeight(r: ReinforcementAttempt): number {
  if (r.analysis.totalCrossings === 0) return r.target.weight;
  return (
    r.target.weight +
    Math.floor(
      (r.energy * r.analysis.mappedCrossings) / r.analysis.totalCrossings
    )
  );
}

/** Is the attempt overfilled? */
export function isOverfilled(r: ReinforcementAttempt): boolean {
  return r.energy > r.target.capacity;
}

/** Energy surplus causing the explosion. */
export function energySurplus(r: ReinforcementAttempt): number {
  return Math.max(0, r.energy - r.target.capacity);
}

/** Wasted energy hitting blind spots. */
export function wastedEnergy(r: ReinforcementAttempt): number {
  if (r.analysis.totalCrossings === 0) return r.energy;
  return Math.floor(
    (r.energy * blindSpots(r.analysis)) / r.analysis.totalCrossings
  );
}

/** Execute a reinforcement attempt. */
export function reinforce(r: ReinforcementAttempt): ReinforcementOutcome {
  // Zero energy or peak conflict → fizzle
  if (r.energy === 0 || r.target.crossings === 0) {
    return { kind: "fizzle" };
  }
  // Overfill → shatter
  if (isOverfilled(r)) {
    return { kind: "shatter" };
  }
  // Success
  return { kind: "success", newWeight: reinforcedWeight(r) };
}

// ---------------------------------------------------------------------------
// Part III: Meander vs Greedy Injection
// ---------------------------------------------------------------------------

/** Per-crossing energy load when meandering through pathways. */
export function perCrossingLoad(energy: number, crossings: number): number {
  if (crossings === 0) return energy; // no pathways, full load
  return Math.floor(energy / crossings);
}

/** Is the meander safe? Per-crossing load within capacity. */
export function isMeanderSafe(
  energy: number,
  crossings: number,
  capacity: number
): boolean {
  return perCrossingLoad(energy, crossings) <= capacity;
}

// ---------------------------------------------------------------------------
// Part IV: The Lightbulb Paradox
// ---------------------------------------------------------------------------

export interface ReinforcedObject {
  baseWeight: number;
  amplification: number;
  lifespanBase: number;
}

/** Reinforced output: base + amplification. */
export function reinforcedOutput(r: ReinforcedObject): number {
  return r.baseWeight + r.amplification;
}

/** Reinforced lifespan: inversely related to amplification. */
export function reinforcedLifespan(r: ReinforcedObject): number {
  return Math.floor(r.lifespanBase / (r.amplification + 1));
}

/** Total output across lifespan. */
export function totalOutput(r: ReinforcedObject): number {
  return reinforcedOutput(r) * reinforcedLifespan(r);
}

// ---------------------------------------------------------------------------
// Part V: Projection vs Reinforcement
// ---------------------------------------------------------------------------

/** Projection: creating topology from nothing. */
export function isProjection(
  crossingsBefore: number,
  crossingsAfter: number
): boolean {
  return crossingsBefore === 0 && crossingsAfter > 0;
}

/** Reinforcement: amplifying weight on existing topology. */
export function isReinforcement(
  crossingsBefore: number,
  crossingsAfter: number,
  weightBefore: number,
  weightAfter: number
): boolean {
  return crossingsBefore === crossingsAfter && weightAfter > weightBefore;
}

// ---------------------------------------------------------------------------
// Part VI: Diagnostic
// ---------------------------------------------------------------------------

export interface ReinforcementDiagnostic {
  /** Can reinforcement succeed? */
  viable: boolean;
  /** Reason if not viable */
  reason?: "peak-conflict" | "overfill" | "zero-energy";
  /** Analysis quality (0-100) */
  analysisAccuracy: number;
  /** Blind spots in the analysis */
  blindSpotCount: number;
  /** Per-crossing energy load */
  perCrossingLoad: number;
  /** Estimated new weight on success */
  estimatedWeight: number;
  /** Estimated lifespan reduction factor */
  lifespanFactor: number;
}

/** Full diagnostic for a reinforcement attempt. */
export function diagnoseReinforcement(
  r: ReinforcementAttempt
): ReinforcementDiagnostic {
  const spots = blindSpots(r.analysis);
  const load = perCrossingLoad(r.energy, r.target.crossings);
  const viable =
    r.energy > 0 && r.target.crossings > 0 && !isOverfilled(r);

  let reason: ReinforcementDiagnostic["reason"];
  if (r.energy === 0) reason = "zero-energy";
  else if (r.target.crossings === 0) reason = "peak-conflict";
  else if (isOverfilled(r)) reason = "overfill";

  const diagnostic: ReinforcementDiagnostic = {
    viable,
    analysisAccuracy: r.analysis.accuracy,
    blindSpotCount: spots,
    perCrossingLoad: load,
    estimatedWeight: reinforcedWeight(r),
    lifespanFactor: r.energy > 0 ? 1 / (r.energy + 1) : 1,
  };
  if (reason !== undefined) {
    diagnostic.reason = reason;
  }
  return diagnostic;
}
