// @a0n/.ignored_gnosis/src/repl-visualization.ts
import type { GraphAST } from './betty/compiler';

const MAX_QUANTUM_EVENTS = 6;
const QUANTUM_EVENT_MARKERS = [
  '[SUPERPOSE]',
  '[EVOLVE]',
  '[ENTANGLE]',
  '[OBSERVE]',
  '[MEASURE]',
  '[HALT]',
  'Race concluded!',
] as const;

export type VisualNodeStatus = 'pending' | 'active' | 'completed' | 'cycled';

export interface ReplVisualizationState {
  nodeStates: Record<string, VisualNodeStatus>;
  activeWaveFunction: string[];
  quantumEvents: string[];
}

function getOrderedNodeIds(ast: GraphAST | null): string[] {
  return ast ? Array.from(ast.nodes.keys()) : [];
}

function parseNodeList(raw: string): string[] {
  return raw
    .split(',')
    .map((segment) => segment.trim())
    .filter((segment) => segment.length > 0);
}

function isQuantumEvent(line: string): boolean {
  return QUANTUM_EVENT_MARKERS.some((marker) => line.includes(marker));
}

function ensureNode(orderedNodeIds: string[], nodeId: string): void {
  if (nodeId.length === 0) {
    return;
  }
  if (!orderedNodeIds.includes(nodeId)) {
    orderedNodeIds.push(nodeId);
  }
}

/**
 * Creates the Initial Visualization.
 */
export function createInitialVisualization(
  ast: GraphAST | null
): ReplVisualizationState {
  const nodeStates: Record<string, VisualNodeStatus> = {};
  for (const nodeId of getOrderedNodeIds(ast)) {
    nodeStates[nodeId] = 'pending';
  }

  return {
    nodeStates,
    activeWaveFunction: [],
    quantumEvents: [],
  };
}

/**
 * Builds the Visualization From Execution.
 */
export function buildVisualizationFromExecution(
  ast: GraphAST | null,
  executionOutput: string
): ReplVisualizationState {
  const orderedNodeIds = getOrderedNodeIds(ast);
  const completedNodes = new Set<string>();
  const activeNodes = new Set<string>();
  const cycledNodes = new Set<string>();
  const quantumEvents: string[] = [];

  for (const rawLine of executionOutput.split('\n')) {
    const line = rawLine.trim();
    if (line.length === 0) {
      continue;
    }

    if (isQuantumEvent(line)) {
      quantumEvents.push(line);
    }

    const rootMatch = line.match(/^Tracing from root:\s*(.+)$/);
    if (rootMatch) {
      const rootNode = rootMatch[1].trim();
      if (rootNode.length > 0 && rootNode !== 'undefined') {
        ensureNode(orderedNodeIds, rootNode);
        activeNodes.add(rootNode);
      }
      continue;
    }

    const executingMatch = line.match(/^-> Executing \[([^\]]+)\]/);
    if (executingMatch) {
      const executedNode = executingMatch[1].trim();
      if (executedNode.length > 0) {
        ensureNode(orderedNodeIds, executedNode);
        completedNodes.add(executedNode);
        activeNodes.delete(executedNode);
      }
      continue;
    }

    const waveMatch = line.match(/Active wave-function:\s*\[([^\]]*)\]/);
    if (waveMatch) {
      activeNodes.clear();
      for (const nodeId of parseNodeList(waveMatch[1])) {
        ensureNode(orderedNodeIds, nodeId);
        activeNodes.add(nodeId);
      }
      continue;
    }

    const raceWinnerMatch = line.match(/Race concluded!\s*Winner:\s*([^\s]+)/);
    if (raceWinnerMatch) {
      const winningNode = raceWinnerMatch[1].trim();
      if (winningNode.length > 0) {
        ensureNode(orderedNodeIds, winningNode);
        activeNodes.clear();
        activeNodes.add(winningNode);
      }
      continue;
    }

    const cycleMatch = line.match(/Cycle detected at ([^.]+)\.?/);
    if (cycleMatch) {
      const cycledNode = cycleMatch[1].trim();
      if (cycledNode.length > 0) {
        ensureNode(orderedNodeIds, cycledNode);
        cycledNodes.add(cycledNode);
      }
    }
  }

  const nodeStates: Record<string, VisualNodeStatus> = {};
  for (const nodeId of orderedNodeIds) {
    nodeStates[nodeId] = 'pending';
  }
  for (const nodeId of completedNodes) {
    nodeStates[nodeId] = 'completed';
  }
  for (const nodeId of activeNodes) {
    nodeStates[nodeId] = 'active';
  }
  for (const nodeId of cycledNodes) {
    nodeStates[nodeId] = 'cycled';
  }

  return {
    nodeStates,
    activeWaveFunction: Array.from(activeNodes),
    quantumEvents: quantumEvents.slice(-MAX_QUANTUM_EVENTS),
  };
}
