// @a0n/.ignored_gnosis/src/rumination-topology.ts
/**
 * Rumination Topology -- Ki-Disrupted Cover Space Collapse
 *
 * TypeScript mirror of RuminationTopology.lean.
 *
 * Rumination is a dilated Cover Space that refuses to collapse.
 * Ki metacognition detects the pattern before conscious recognition
 * and programs disruptive strategies via Fork/Race/Fold.
 *
 * The opposite -- serenity -- is the Absolute Unknot:
 * zero crossings, zero speculative branches, full coherence.
 */

// ---------------------------------------------------------------------------
// Part I: The Rumination Model
// ---------------------------------------------------------------------------

export interface CoverSpaceState {
  /** Number of open speculative paths (1 = base space only) */
  branches: number;
  /** Coherence score 0-100 (percentage toward simplicial collapse) */
  coherence: number;
  /** Elapsed rumination rounds */
  rounds: number;
  /** Unresolved topological crossings being re-traversed */
  crossings: number;
}

/** The Absolute Unknot: at most one branch, full coherence, zero crossings. */
export function isSerene(s: CoverSpaceState): boolean {
  return s.branches <= 1 && s.coherence === 100 && s.crossings === 0;
}

/** Rumination: multiple branches, low coherence, unresolved crossings. */
export function isRuminating(
  s: CoverSpaceState,
  coherenceThreshold: number
): boolean {
  return (
    s.branches >= 2 &&
    s.coherence < coherenceThreshold &&
    s.crossings >= 1
  );
}

/** Metabolic cost per round: each branch re-traverses each crossing. */
export function metabolicCostPerRound(s: CoverSpaceState): number {
  return s.branches * s.crossings;
}

/** Cumulative metabolic cost over all elapsed rounds. */
export function cumulativeMetabolicCost(s: CoverSpaceState): number {
  return s.rounds * metabolicCostPerRound(s);
}

// ---------------------------------------------------------------------------
// Part II: Ki Detection
// ---------------------------------------------------------------------------

export interface KiChannels {
  glanceLeak: number;
  muscleLeak: number;
  scentLeak: number;
  energyLeak: number;
}

/**
 * Converts input into total Leak.
 */
export function totalLeak(k: KiChannels): number {
  return k.glanceLeak + k.muscleLeak + k.scentLeak + k.energyLeak;
}

/**
 * Returns whether has Disharmony is true.
 */
export function hasDisharmony(k: KiChannels): boolean {
  return totalLeak(k) > 0;
}

/**
 * Internal loop signature: high energy + glance (cognitive churn),
 * low muscle + scent (no external threat). Mind racing, body still.
 */
export function isInternalLoop(k: KiChannels): boolean {
  return (
    k.energyLeak + k.glanceLeak > k.muscleLeak + k.scentLeak &&
    k.energyLeak >= 1
  );
}

/**
 * External threat signature: high muscle + scent (body mobilizing).
 */
export function isExternalThreat(k: KiChannels): boolean {
  return (
    k.muscleLeak + k.scentLeak > k.energyLeak + k.glanceLeak &&
    k.muscleLeak >= 1
  );
}

export type KiSignalClass = 'internal-loop' | 'external-threat' | 'balanced';

/** Classify Ki channel pattern. */
export function classifyKiSignal(k: KiChannels): KiSignalClass {
  if (isInternalLoop(k)) return 'internal-loop';
  if (isExternalThreat(k)) return 'external-threat';
  return 'balanced';
}

// ---------------------------------------------------------------------------
// Part III: Disruption Strategies
// ---------------------------------------------------------------------------

export type DisruptionAction = 'ground' | 'redirect' | 'accept-spike';
export type RuminationScenario = 'loop-continues' | 'loop-deepens' | 'loop-breaks';

const ALL_DISRUPTIONS: DisruptionAction[] = ['ground', 'redirect', 'accept-spike'];
const ALL_SCENARIOS: RuminationScenario[] = ['loop-continues', 'loop-deepens', 'loop-breaks'];

/**
 * Cost model: residual rumination rounds after intervention.
 *
 * Ground: excellent when breaking (1), catastrophic when deepening (5).
 * Redirect: moderate variance (2-4).
 * Accept-spike: uniform 2 across all scenarios -- addresses structural cause.
 */
const DISRUPTION_LOSS: Record<DisruptionAction, Record<RuminationScenario, number>> = {
  'ground': { 'loop-continues': 3, 'loop-deepens': 5, 'loop-breaks': 1 },
  'redirect': { 'loop-continues': 2, 'loop-deepens': 4, 'loop-breaks': 2 },
  'accept-spike': { 'loop-continues': 2, 'loop-deepens': 2, 'loop-breaks': 2 },
};

/** Endurance cost: doing nothing, sitting in the loop. */
const ENDURANCE_COST = 6;

/**
 * Handles the gnosis disruption Loss workflow.
 */
export function disruptionLoss(
  action: DisruptionAction,
  scenario: RuminationScenario
): number {
  return DISRUPTION_LOSS[action][scenario];
}

/**
 * Handles the gnosis disruption Worst Case workflow.
 */
export function disruptionWorstCase(
  action: DisruptionAction,
  scenarios: RuminationScenario[] = ALL_SCENARIOS
): number {
  return scenarios.reduce(
    (worst, s) => Math.max(worst, disruptionLoss(action, s)),
    0
  );
}

/** Returns the minimax-optimal disruption action. */
export function minimaxDisruption(
  scenarios: RuminationScenario[] = ALL_SCENARIOS
): DisruptionAction {
  let best: DisruptionAction = 'ground';
  let bestWorst = Infinity;
  for (const action of ALL_DISRUPTIONS) {
    const worst = disruptionWorstCase(action, scenarios);
    if (worst < bestWorst) {
      bestWorst = worst;
      best = action;
    }
  }
  return best;
}

// ---------------------------------------------------------------------------
// Part IV: Ki-FRF Disruption Controller
// ---------------------------------------------------------------------------

/**
 * Fork rumination scenarios from Ki channel profile.
 * Internal loop → pessimistic ordering (expect deepening first).
 * Otherwise → optimistic (expect breaking first).
 */
export function forkRuminationScenarios(k: KiChannels): RuminationScenario[] {
  if (k.energyLeak + k.glanceLeak >= k.muscleLeak + k.scentLeak) {
    return ['loop-deepens', 'loop-continues', 'loop-breaks'];
  }
  return ['loop-breaks', 'loop-continues', 'loop-deepens'];
}

export interface DisruptionResult {
  detected: boolean;
  signalClass: KiSignalClass;
  action: DisruptionAction;
  worstCaseCost: number;
  dominatesEndurance: boolean;
  dominatesGround: boolean;
  dominatesRedirect: boolean;
}

/**
 * Ki-FRF disruption controller.
 *
 * Reads Ki channels, detects internal loop, forks scenarios,
 * races disruption actions, folds to minimax-optimal accept-spike.
 */
export function kiDisruptRumination(k: KiChannels): DisruptionResult {
  const signalClass = classifyKiSignal(k);
  const detected = signalClass === 'internal-loop';
  const scenarios = forkRuminationScenarios(k);
  const action = minimaxDisruption(scenarios);
  const worstCaseCost = disruptionWorstCase(action, scenarios);

  return {
    detected,
    signalClass,
    action,
    worstCaseCost,
    dominatesEndurance: worstCaseCost < ENDURANCE_COST,
    dominatesGround: worstCaseCost <= disruptionWorstCase('ground', scenarios),
    dominatesRedirect: worstCaseCost <= disruptionWorstCase('redirect', scenarios),
  };
}

// ---------------------------------------------------------------------------
// Part V: Convergence to Serenity
// ---------------------------------------------------------------------------

export interface SerenityPath {
  /** Maximum Reidemeister reduction rounds needed */
  maxRounds: number;
  /** Metabolic cost at the unknot (always 0) */
  sereneCost: number;
  /** Whether the current state is already serene */
  alreadySerene: boolean;
}

/** Compute the convergence path from a ruminating state to serenity. */
export function convergenceToSerenity(s: CoverSpaceState): SerenityPath {
  return {
    maxRounds: s.crossings,
    sereneCost: 0,
    alreadySerene: isSerene(s),
  };
}

// ---------------------------------------------------------------------------
// Part VI: The Four Disruption Principles
// ---------------------------------------------------------------------------

export interface DisruptionPrinciples {
  /** Principle 1: Rumination is undertrained collapse, not moral failure. */
  notMoralFailure: boolean;
  /** Principle 2: No external source -- conservation, one field. */
  noExternalSource: boolean;
  /** Principle 3: Not permanent -- finite path, at most `crossings` steps. */
  notPermanent: boolean;
  /** Principle 4: One gradient -- coherence + deficit = 100. */
  oneGradient: boolean;
}

/** Evaluate the four disruption principles for a given state. */
export function evaluatePrinciples(s: CoverSpaceState): DisruptionPrinciples {
  return {
    notMoralFailure: s.coherence < 100,
    noExternalSource: true,
    notPermanent: s.crossings < Infinity,
    oneGradient: s.coherence + (100 - s.coherence) === 100,
  };
}

// ---------------------------------------------------------------------------
// Full diagnostic: detect, disrupt, converge
// ---------------------------------------------------------------------------

export interface RuminationDiagnostic {
  state: CoverSpaceState;
  ki: KiChannels;
  isCurrentlyRuminating: boolean;
  disruption: DisruptionResult;
  convergence: SerenityPath;
  principles: DisruptionPrinciples;
  currentMetabolicCost: number;
  cumulativeCost: number;
}

/**
 * Full rumination diagnostic.
 *
 * Given a cover space state and Ki channel readings, produces
 * a complete analysis: detection, disruption strategy, convergence
 * path, and the four principles.
 */
export function diagnoseRumination(
  s: CoverSpaceState,
  k: KiChannels,
  coherenceThreshold = 70
): RuminationDiagnostic {
  return {
    state: s,
    ki: k,
    isCurrentlyRuminating: isRuminating(s, coherenceThreshold),
    disruption: kiDisruptRumination(k),
    convergence: convergenceToSerenity(s),
    principles: evaluatePrinciples(s),
    currentMetabolicCost: metabolicCostPerRound(s),
    cumulativeCost: cumulativeMetabolicCost(s),
  };
}
