// @a0n/.ignored_gnosis/src/scaffold-test.ts
import * as path from 'node:path';
import * as fs from 'node:fs';
import {
  generateBranchTests,
  type TargetFunction,
} from '@a0n/polyglot-scanner-core';
import { extractTopology } from '../../a0/src/topology-test-gen';

export async function runScaffoldTest(filePath: string): Promise<number> {
  const absolutePath = path.resolve(filePath);
  if (!fs.existsSync(absolutePath)) {
    console.error(`Error: File not found: ${absolutePath}`);
    return 1;
  }

  const source = fs.readFileSync(absolutePath, 'utf8');
  const topologies = extractTopology(source, absolutePath);
  
  if (!topologies || topologies.length === 0) {
    console.log('No function topologies found in the file.');
    return 0;
  }

  const exportedTopologies = topologies.filter(t => t.exported);
  if (exportedTopologies.length === 0) {
    console.log('No exported function topologies found in the file.');
    return 0;
  }

  const baseName = path.basename(absolutePath, path.extname(absolutePath));
  console.log(`Found ${exportedTopologies.length} exported functions. Generating tests...`);

  let totalTests = 0;
  for (const topology of exportedTopologies) {
    if (topology.branches.length === 0) {
      console.log(`Skipping ${topology.name}: no branches found.`);
      continue;
    }

    const target: TargetFunction = {
      name: topology.name,
      filePath: absolutePath,
      source: topology.jsSource,
      importName: topology.name,
      importPath: `./${path.basename(absolutePath)}`,
      params: topology.params.map(p => ({ name: p.name, type: p.type })),
      branches: topology.branches.map(b => ({
        index: b.branchIndex,
        condition: b.condition,
        testsParam: b.testsParam,
      })),
      isAsync: topology.async,
    };

    console.log(`  Targeting ${topology.name}...`);
    
    try {
      const tests = await generateBranchTests([target], {
        tier: 'topology',
        maxGenerations: 2,
      });

      if (tests.length === 0) {
        console.log(`    No tests were generated for ${topology.name}.`);
        continue;
      }

      const testDir = path.join(path.dirname(absolutePath), '__tests__');
      if (!fs.existsSync(testDir)) {
        fs.mkdirSync(testDir, { recursive: true });
      }

      for (const test of tests) {
        const testPath = path.join(testDir, `${baseName}.${topology.name.toLowerCase()}.test.ts`);
        fs.writeFileSync(testPath, test.testCode, 'utf8');
        console.log(`    Wrote test file to ${testPath}`);
        totalTests++;
      }
    } catch (error) {
      console.error(`    Test generation failed for ${topology.name}: ${error}`);
    }
  }

  return 0;
}
