// @a0n/.ignored_gnosis/src/seismic-stress-toy.ts
/**
 * **Not verified by Lean.** Discrete-time toy for **load → threshold → slip** on a single fault patch
 * or asperity (stress accumulation, rupture when a critical value is crossed, stress release).
 *
 * This is **not** elastodynamics, rate-and-state friction, or rupture nucleation on a continuum fault —
 * it is a **deterministic finite-state update** you can compose into larger simulators. For formal
 * discrete mirrors see `PlanetaryHomologySandbox.SeismicStressToySketch` in Lean (`stepStress` /
 * `stepStressSubtract`). `stepCoupledTwoPatch` / `simulateCoupledTwoPatch` / `stepCoupledTwoPatchRateState` /
 * `simulateCoupledTwoPatchRateState` add a **two-node** relaxation split before slip (conserving total
 * loaded stress on the relax step) — a graph toy, not a Green’s function
 * for elastic half-space. Optional **`thetaOnRupture`** resets θ after slip in rate-state stepping.
 *
 * @module @a0n/gnosis/seismic-stress-toy
 */

/** Explicit stance for manifests and UIs. */
export const SEISMIC_STRESS_TOY_DISCLAIMER =
  'This module is a deterministic discrete-time stress/threshold toy; it is not a continuum earthquake model.';

/** How stress is set immediately after a rupture. */
export type SeismicToyRuptureMode = 'subtract' | 'reset';

/** Parameters for one asperity (single patch). All numeric fields must be finite; rates are per step. */
export interface SingleAsperityParams {
  /** Shear (or Coulomb) stress at which slip occurs after loading. Must be positive. */
  readonly criticalStress: number;
  /** Stress added each discrete step before the threshold check. Non-negative. */
  readonly loadingRate: number;
  /** Stress drop on slip when `mode === 'subtract'` (ignored for `'reset'`). Non-negative. */
  readonly stressDrop: number;
  /** Floor after slip (`subtract`) or fixed post-slip value (`reset`). Non-negative. */
  readonly residualStress: number;
  readonly mode: SeismicToyRuptureMode;
}

export interface StressStepResult {
  readonly stress: number;
  /** Present when a rupture occurred this step. */
  readonly rupture: SeismicRuptureEvent | null;
}

export interface SeismicRuptureEvent {
  /** Stress immediately before slip (after loading). */
  readonly stressBeforeSlip: number;
  /** Stress immediately after applying the rupture rule. */
  readonly stressAfterSlip: number;
}

export interface MultiAsperityParams {
  readonly patches: readonly SingleAsperityParams[];
}

function assertFiniteNonNeg(name: string, x: number): void {
  if (!Number.isFinite(x) || x < 0) {
    throw new RangeError(`${name} must be a finite number ≥ 0`);
  }
}

function assertFinitePositive(name: string, x: number): void {
  if (!Number.isFinite(x) || x <= 0) {
    throw new RangeError(`${name} must be a finite number > 0`);
  }
}

/**
 * Validate and return a normalized params object (defensive copy).
 */
export function normalizeSingleAsperityParams(p: SingleAsperityParams): SingleAsperityParams {
  assertFinitePositive('criticalStress', p.criticalStress);
  assertFiniteNonNeg('loadingRate', p.loadingRate);
  assertFiniteNonNeg('stressDrop', p.stressDrop);
  assertFiniteNonNeg('residualStress', p.residualStress);
  if (p.mode !== 'subtract' && p.mode !== 'reset') {
    throw new RangeError("mode must be 'subtract' or 'reset'");
  }
  return {
    criticalStress: p.criticalStress,
    loadingRate: p.loadingRate,
    stressDrop: p.stressDrop,
    residualStress: p.residualStress,
    mode: p.mode,
  };
}

/**
 * Rupture rule on **already loaded** shear stress (does not add `loadingRate`).
 * Use after coupling or other sub-steps; mirrors Lean `stepStressSubtract` / `stepStress`.
 */
export function slipAfterLoaded(loadedStress: number, params: SingleAsperityParams): StressStepResult {
  if (!Number.isFinite(loadedStress)) {
    throw new RangeError('loadedStress must be finite');
  }
  const p = normalizeSingleAsperityParams(params);
  if (loadedStress < p.criticalStress) {
    return { stress: loadedStress, rupture: null };
  }
  let after: number;
  if (p.mode === 'reset') {
    after = p.residualStress;
  } else {
    after = Math.max(p.residualStress, loadedStress - p.stressDrop);
  }
  return {
    stress: after,
    rupture: { stressBeforeSlip: loadedStress, stressAfterSlip: after },
  };
}

/**
 * One discrete step: add loading, then if `stress >= criticalStress` apply the rupture rule.
 */
export function stepSingleAsperity(
  stress: number,
  params: SingleAsperityParams,
): StressStepResult {
  if (!Number.isFinite(stress)) {
    throw new RangeError('stress must be finite');
  }
  const p = normalizeSingleAsperityParams(params);
  return slipAfterLoaded(stress + p.loadingRate, p);
}

/**
 * **Rate–state-shaped hook only:** θ ← θ + `agingRate`·Δt — not the Dieterich–Ruina ODE system.
 * Pair with `effectiveCriticalLinear` to vary a scalar strength proxy between ruptures.
 */
export function rateStateThetaAging(theta: number, dt: number, agingRate: number): number {
  if (!Number.isFinite(theta) || !Number.isFinite(dt) || !Number.isFinite(agingRate)) {
    throw new RangeError('theta, dt, and agingRate must be finite');
  }
  if (dt < 0) {
    throw new RangeError('dt must be non-negative');
  }
  return theta + agingRate * dt;
}

/**
 * Linear state-dependent critical stress μ* = μ₀ + b·θ (arbitrary units; sign of `b` is modeler choice).
 */
export function effectiveCriticalLinear(critical0: number, b: number, theta: number): number {
  if (!Number.isFinite(critical0) || !Number.isFinite(b) || !Number.isFinite(theta)) {
    throw new RangeError('critical0, b, and theta must be finite');
  }
  return critical0 + b * theta;
}

/**
 * Single asperity with **linear** μ*(θ) = μ₀ + b·θ and aging at the **start** of each step (then load +
 * rupture). θ is updated before `criticalStress` is applied — a discrete skeleton, not a coupled ODE.
 */
export interface RateStateAsperityParams {
  readonly loadingRate: number;
  readonly stressDrop: number;
  readonly residualStress: number;
  readonly mode: SeismicToyRuptureMode;
  /** Baseline strength proxy (must yield positive `criticalStress` with your θ range). */
  readonly critical0: number;
  /** Strength sensitivity to θ (sign is modeler choice). */
  readonly b: number;
  readonly agingRate: number;
  readonly dt: number;
  /**
   * If set, when a rupture occurs this step θ is set to this value (e.g. `0`) instead of the post-aging θ.
   */
  readonly thetaOnRupture?: number;
}

function resolveThetaAfterStep(
  thetaNext: number,
  rupture: SeismicRuptureEvent | null,
  thetaOnRupture: number | undefined,
): number {
  if (rupture !== null && thetaOnRupture !== undefined) {
    return thetaOnRupture;
  }
  return thetaNext;
}

export interface RateStateStepResult extends StressStepResult {
  /** θ after this step (possibly reset on rupture; see `thetaOnRupture`). */
  readonly theta: number;
}

/**
 * Handles the gnosis normalize Rate State Asperity Params workflow.
 */
export function normalizeRateStateAsperityParams(p: RateStateAsperityParams): RateStateAsperityParams {
  assertFinitePositive('critical0', p.critical0);
  assertFiniteNonNeg('loadingRate', p.loadingRate);
  assertFiniteNonNeg('stressDrop', p.stressDrop);
  assertFiniteNonNeg('residualStress', p.residualStress);
  assertFiniteNonNeg('dt', p.dt);
  if (!Number.isFinite(p.b) || !Number.isFinite(p.agingRate)) {
    throw new RangeError('b and agingRate must be finite');
  }
  if (p.mode !== 'subtract' && p.mode !== 'reset') {
    throw new RangeError("mode must be 'subtract' or 'reset'");
  }
  if (p.thetaOnRupture !== undefined && !Number.isFinite(p.thetaOnRupture)) {
    throw new RangeError('thetaOnRupture must be finite when provided');
  }
  return {
    loadingRate: p.loadingRate,
    stressDrop: p.stressDrop,
    residualStress: p.residualStress,
    mode: p.mode,
    critical0: p.critical0,
    b: p.b,
    agingRate: p.agingRate,
    dt: p.dt,
    ...(p.thetaOnRupture !== undefined
      ? { thetaOnRupture: p.thetaOnRupture }
      : {}),
  };
}

/**
 * Age θ → build μ* → `stepSingleAsperity`. Fails if computed `criticalStress` is not positive.
 */
export function stepSingleAsperityRateState(
  stress: number,
  theta: number,
  params: RateStateAsperityParams,
): RateStateStepResult {
  if (!Number.isFinite(stress) || !Number.isFinite(theta)) {
    throw new RangeError('stress and theta must be finite');
  }
  const pb = normalizeRateStateAsperityParams(params);
  const thetaNext = rateStateThetaAging(theta, pb.dt, pb.agingRate);
  const critical = effectiveCriticalLinear(pb.critical0, pb.b, thetaNext);
  const sp: SingleAsperityParams = {
    criticalStress: critical,
    loadingRate: pb.loadingRate,
    stressDrop: pb.stressDrop,
    residualStress: pb.residualStress,
    mode: pb.mode,
  };
  const r = stepSingleAsperity(stress, sp);
  const thetaOut = resolveThetaAfterStep(thetaNext, r.rupture, pb.thetaOnRupture);
  return { ...r, theta: thetaOut };
}

/**
 * Trajectory with θ tracked (length `steps + 1` entries for σ and θ).
 */
export function simulateSingleAsperityRateState(
  initialStress: number,
  initialTheta: number,
  params: RateStateAsperityParams,
  steps: number,
): {
  readonly stresses: readonly number[];
  readonly thetas: readonly number[];
  readonly ruptures: readonly (SeismicRuptureEvent | null)[];
} {
  if (!Number.isInteger(steps) || steps < 0) {
    throw new RangeError('steps must be a non-negative integer');
  }
  if (!Number.isFinite(initialTheta)) {
    throw new RangeError('initialTheta must be finite');
  }
  const stresses: number[] = [initialStress];
  const thetas: number[] = [initialTheta];
  const ruptures: (SeismicRuptureEvent | null)[] = [];
  let σ = initialStress;
  let θ = initialTheta;
  for (let i = 0; i < steps; i++) {
    const r = stepSingleAsperityRateState(σ, θ, params);
    σ = r.stress;
    θ = r.theta;
    stresses.push(σ);
    thetas.push(θ);
    ruptures.push(r.rupture);
  }
  return { stresses, thetas, ruptures };
}

/**
 * Two-patch symmetric relaxation: moves each toward the other (conserves `σ₀ + σ₁`).
 * `κ` is in **`[0, 1]`**, matching Lean `relax_two_patch_real_conserves` / `relax_two_patch_parts_nonneg_of_unit_kappa`
 * (for nonnegative **loaded** stresses, both outputs stay nonnegative).
 */
export function relaxTwoPatch(
  loaded0: number,
  loaded1: number,
  kappa: number,
): { readonly m0: number; readonly m1: number } {
  if (!Number.isFinite(loaded0) || !Number.isFinite(loaded1)) {
    throw new RangeError('stresses must be finite');
  }
  if (!Number.isFinite(kappa) || kappa < 0 || kappa > 1) {
    throw new RangeError('kappa must be finite and in [0, 1]');
  }
  const d = loaded1 - loaded0;
  return {
    m0: loaded0 + kappa * d,
    m1: loaded1 - kappa * d,
  };
}

export interface CoupledTwoPatchParams {
  readonly kappa: number;
  readonly patch0: SingleAsperityParams;
  readonly patch1: SingleAsperityParams;
}

export interface CoupledTwoPatchRateStateParams {
  readonly kappa: number;
  readonly patch0: RateStateAsperityParams;
  readonly patch1: RateStateAsperityParams;
}

/**
 * Age each θ → μ* → load → `relaxTwoPatch` → slip. Independent aging timesteps per patch.
 */
export function stepCoupledTwoPatchRateState(
  stress0: number,
  stress1: number,
  theta0: number,
  theta1: number,
  coupled: CoupledTwoPatchRateStateParams,
): {
  readonly stress0: number;
  readonly stress1: number;
  readonly theta0: number;
  readonly theta1: number;
  readonly rupture0: SeismicRuptureEvent | null;
  readonly rupture1: SeismicRuptureEvent | null;
} {
  if (!Number.isFinite(stress0) || !Number.isFinite(stress1) || !Number.isFinite(theta0) || !Number.isFinite(theta1)) {
    throw new RangeError('stress and theta values must be finite');
  }
  const p0b = normalizeRateStateAsperityParams(coupled.patch0);
  const p1b = normalizeRateStateAsperityParams(coupled.patch1);
  const t0 = rateStateThetaAging(theta0, p0b.dt, p0b.agingRate);
  const t1 = rateStateThetaAging(theta1, p1b.dt, p1b.agingRate);
  const crit0 = effectiveCriticalLinear(p0b.critical0, p0b.b, t0);
  const crit1 = effectiveCriticalLinear(p1b.critical0, p1b.b, t1);
  const sp0: SingleAsperityParams = {
    criticalStress: crit0,
    loadingRate: p0b.loadingRate,
    stressDrop: p0b.stressDrop,
    residualStress: p0b.residualStress,
    mode: p0b.mode,
  };
  const sp1: SingleAsperityParams = {
    criticalStress: crit1,
    loadingRate: p1b.loadingRate,
    stressDrop: p1b.stressDrop,
    residualStress: p1b.residualStress,
    mode: p1b.mode,
  };
  const l0 = stress0 + p0b.loadingRate;
  const l1 = stress1 + p1b.loadingRate;
  const { m0, m1 } = relaxTwoPatch(l0, l1, coupled.kappa);
  const r0 = slipAfterLoaded(m0, sp0);
  const r1 = slipAfterLoaded(m1, sp1);
  const outTheta0 = resolveThetaAfterStep(t0, r0.rupture, p0b.thetaOnRupture);
  const outTheta1 = resolveThetaAfterStep(t1, r1.rupture, p1b.thetaOnRupture);
  return {
    stress0: r0.stress,
    stress1: r1.stress,
    theta0: outTheta0,
    theta1: outTheta1,
    rupture0: r0.rupture,
    rupture1: r1.rupture,
  };
}

/**
 * Multi-step trajectory for `stepCoupledTwoPatchRateState` (length `steps + 1` samples per series).
 */
export function simulateCoupledTwoPatchRateState(
  initialStress0: number,
  initialStress1: number,
  initialTheta0: number,
  initialTheta1: number,
  coupled: CoupledTwoPatchRateStateParams,
  steps: number,
): {
  readonly stresses0: readonly number[];
  readonly stresses1: readonly number[];
  readonly thetas0: readonly number[];
  readonly thetas1: readonly number[];
  readonly ruptures0: readonly (SeismicRuptureEvent | null)[];
  readonly ruptures1: readonly (SeismicRuptureEvent | null)[];
} {
  if (!Number.isInteger(steps) || steps < 0) {
    throw new RangeError('steps must be a non-negative integer');
  }
  if (
    !Number.isFinite(initialStress0) ||
    !Number.isFinite(initialStress1) ||
    !Number.isFinite(initialTheta0) ||
    !Number.isFinite(initialTheta1)
  ) {
    throw new RangeError('initial stresses and thetas must be finite');
  }
  const s0: number[] = [initialStress0];
  const s1: number[] = [initialStress1];
  const t0: number[] = [initialTheta0];
  const t1: number[] = [initialTheta1];
  const r0: (SeismicRuptureEvent | null)[] = [];
  const r1: (SeismicRuptureEvent | null)[] = [];
  let σ0 = initialStress0;
  let σ1 = initialStress1;
  let θ0 = initialTheta0;
  let θ1 = initialTheta1;
  for (let i = 0; i < steps; i++) {
    const step = stepCoupledTwoPatchRateState(σ0, σ1, θ0, θ1, coupled);
    σ0 = step.stress0;
    σ1 = step.stress1;
    θ0 = step.theta0;
    θ1 = step.theta1;
    s0.push(σ0);
    s1.push(σ1);
    t0.push(θ0);
    t1.push(θ1);
    r0.push(step.rupture0);
    r1.push(step.rupture1);
  }
  return {
    stresses0: s0,
    stresses1: s1,
    thetas0: t0,
    thetas1: t1,
    ruptures0: r0,
    ruptures1: r1,
  };
}

/**
 * Load both patches, apply symmetric stress transfer (`relaxTwoPatch`), then slip each patch.
 * Order: load → relax → rupture (see module docstring — not a continuum model).
 */
export function stepCoupledTwoPatch(
  stress0: number,
  stress1: number,
  coupled: CoupledTwoPatchParams,
): {
  readonly stress0: number;
  readonly stress1: number;
  readonly rupture0: SeismicRuptureEvent | null;
  readonly rupture1: SeismicRuptureEvent | null;
} {
  const p0 = normalizeSingleAsperityParams(coupled.patch0);
  const p1 = normalizeSingleAsperityParams(coupled.patch1);
  const l0 = stress0 + p0.loadingRate;
  const l1 = stress1 + p1.loadingRate;
  const { m0, m1 } = relaxTwoPatch(l0, l1, coupled.kappa);
  const r0 = slipAfterLoaded(m0, p0);
  const r1 = slipAfterLoaded(m1, p1);
  return {
    stress0: r0.stress,
    stress1: r1.stress,
    rupture0: r0.rupture,
    rupture1: r1.rupture,
  };
}

/**
 * Multi-step trajectory for `stepCoupledTwoPatch` (length `steps + 1` samples per stress series).
 */
export function simulateCoupledTwoPatch(
  initialStress0: number,
  initialStress1: number,
  coupled: CoupledTwoPatchParams,
  steps: number,
): {
  readonly stresses0: readonly number[];
  readonly stresses1: readonly number[];
  readonly ruptures0: readonly (SeismicRuptureEvent | null)[];
  readonly ruptures1: readonly (SeismicRuptureEvent | null)[];
} {
  if (!Number.isInteger(steps) || steps < 0) {
    throw new RangeError('steps must be a non-negative integer');
  }
  if (!Number.isFinite(initialStress0) || !Number.isFinite(initialStress1)) {
    throw new RangeError('initial stresses must be finite');
  }
  const s0: number[] = [initialStress0];
  const s1: number[] = [initialStress1];
  const r0: (SeismicRuptureEvent | null)[] = [];
  const r1: (SeismicRuptureEvent | null)[] = [];
  let σ0 = initialStress0;
  let σ1 = initialStress1;
  for (let i = 0; i < steps; i++) {
    const step = stepCoupledTwoPatch(σ0, σ1, coupled);
    σ0 = step.stress0;
    σ1 = step.stress1;
    s0.push(σ0);
    s1.push(σ1);
    r0.push(step.rupture0);
    r1.push(step.rupture1);
  }
  return {
    stresses0: s0,
    stresses1: s1,
    ruptures0: r0,
    ruptures1: r1,
  };
}

/**
 * Run `steps` updates from `initialStress`, returning the trajectory (length `steps + 1` states).
 */
export function simulateSingleAsperity(
  initialStress: number,
  params: SingleAsperityParams,
  steps: number,
): { readonly stresses: readonly number[]; readonly ruptures: readonly (SeismicRuptureEvent | null)[] } {
  if (!Number.isInteger(steps) || steps < 0) {
    throw new RangeError('steps must be a non-negative integer');
  }
  const stresses: number[] = [initialStress];
  const ruptures: (SeismicRuptureEvent | null)[] = [];
  let σ = initialStress;
  for (let i = 0; i < steps; i++) {
    const r = stepSingleAsperity(σ, params);
    σ = r.stress;
    stresses.push(σ);
    ruptures.push(r.rupture);
  }
  return { stresses, ruptures };
}

/**
 * **Fork/race toy:** independent patches loaded in parallel; each patch may rupture in the same step.
 * Processing order is by index (deterministic tie-break).
 */
export function stepMultiAsperity(
  stresses: readonly number[],
  multi: MultiAsperityParams,
): { readonly stresses: readonly number[]; readonly ruptures: readonly (SeismicRuptureEvent | null)[] } {
  if (stresses.length !== multi.patches.length) {
    throw new RangeError('stresses and patches must have the same length');
  }
  const next: number[] = [];
  const ruptures: (SeismicRuptureEvent | null)[] = [];
  for (let i = 0; i < stresses.length; i++) {
    const r = stepSingleAsperity(stresses[i]!, multi.patches[i]!);
    next.push(r.stress);
    ruptures.push(r.rupture);
  }
  return { stresses: next, ruptures };
}
