// @a0n/.ignored_gnosis/src/self-reference.ts
/**
 * self-reference.ts -- Fixed-point and self-hosting as void walking
 *
 * The identification:
 *   - Gnosis compiling itself (betti.gg) = fixed point of the void walk
 *   - Y combinator = trace in the traced monoidal category
 *   - Quine = topology whose complement distribution reproduces itself
 *   - Gödel numbering = VoidBoundary encoding of topologies
 *   - Diagonal argument = complement distribution over the boundary's own structure
 *
 * The formal ledger entry: self-reference formalizes the void walking the void.
 */

import {
  type VoidBoundary,
  createVoidBoundary,
  updateVoidBoundary,
  complementDistribution,
  shannonEntropy,
} from './void';

// ============================================================================
// Fixed-point structures
// ============================================================================

/**
 * A fixed point of a function f: f(x) = x.
 * In void walking, a fixed point is a boundary whose complement
 * distribution, when used to generate rejections, reproduces itself.
 */
export interface FixedPoint<T> {
  value: T;
  /** Number of iterations to converge */
  iterations: number;
  /** Whether convergence was achieved */
  converged: boolean;
  /** Residual: ||f(x) - x|| at termination */
  residual: number;
}

/**
 * Find a fixed point of a function on VoidBoundaries.
 * Iterate f until ||f(x) - x|| < tolerance or maxIter reached.
 */
export function findFixedPoint(
  initial: VoidBoundary,
  f: (b: VoidBoundary) => VoidBoundary,
  tolerance: number = 1e-8,
  maxIter: number = 1000
): FixedPoint<VoidBoundary> {
  let current = initial;
  for (let i = 0; i < maxIter; i++) {
    const next = f(current);
    const residual = boundaryDistance(current, next);
    if (residual < tolerance) {
      return { value: next, iterations: i + 1, converged: true, residual };
    }
    current = next;
  }
  return {
    value: current,
    iterations: maxIter,
    converged: false,
    residual: boundaryDistance(current, f(current)),
  };
}

function boundaryDistance(a: VoidBoundary, b: VoidBoundary): number {
  const n = Math.min(a.counts.length, b.counts.length);
  let dist = 0;
  for (let i = 0; i < n; i++) {
    dist += (a.counts[i] - b.counts[i]) ** 2;
  }
  return Math.sqrt(dist);
}

// ============================================================================
// Self-referential void boundary
// ============================================================================

/**
 * A self-referential boundary: its complement distribution encodes
 * the boundary's own structure. The complement formalizes the boundary.
 *
 * This is the void walking the void: the boundary measures itself.
 */
export function selfReferentialBoundary(
  n: number,
  eta: number = 3.0,
  maxIter: number = 100
): FixedPoint<VoidBoundary> {
  const initial = createVoidBoundary(n);
  // Seed with small random values
  for (let i = 0; i < n; i++) {
    initial.counts[i] = (i + 1) * 0.1;
  }
  initial.totalEntries = initial.counts.reduce((a, b) => a + b, 0);

  return findFixedPoint(
    initial,
    (b) => {
      const dist = complementDistribution(b, eta);
      // Use complement distribution as new counts (scaled)
      const scale = b.totalEntries > 0 ? b.totalEntries : 1;
      const newCounts = dist.map((p) => p * scale);
      return {
        counts: newCounts,
        totalEntries: newCounts.reduce((a, b) => a + b, 0),
      };
    },
    1e-8,
    maxIter
  );
}

// ============================================================================
// Quine boundary -- reproduces itself through complement
// ============================================================================

/**
 * Check if a boundary is a quine: its complement distribution,
 * when normalized to the same total, equals itself.
 */
export function isQuineBoundary(
  boundary: VoidBoundary,
  eta: number,
  tolerance: number = 1e-6
): boolean {
  const dist = complementDistribution(boundary, eta);
  const total = boundary.totalEntries;
  if (total === 0) return false;

  const normalized = boundary.counts.map((c) => c / total);
  return normalized.every((n, i) => Math.abs(n - dist[i]) < tolerance);
}

/**
 * The uniform boundary maps to a quine (for any eta).
 * When all counts are equal, complement is uniform,
 * and uniform normalized = uniform.
 */
export function uniformQuine(n: number): VoidBoundary {
  const boundary = createVoidBoundary(n);
  for (let i = 0; i < n; i++) {
    boundary.counts[i] = 1;
  }
  boundary.totalEntries = n;
  return boundary;
}

// ============================================================================
// Diagonal construction -- Cantor's argument in void space
// ============================================================================

/**
 * The diagonal boundary: a boundary that differs from every row
 * of a "table" of boundaries. This is Cantor's diagonal argument
 * applied to void space.
 *
 * Given a list of boundaries (all same dimension n), construct
 * a boundary that differs from boundary i at dimension i.
 */
export function diagonalBoundary(boundaries: VoidBoundary[]): VoidBoundary {
  const n = boundaries.length;
  const diagonal = createVoidBoundary(n);
  for (let i = 0; i < n && i < (boundaries[i]?.counts.length ?? 0); i++) {
    // Differ from boundary i at position i
    diagonal.counts[i] = boundaries[i].counts[i] + 1;
  }
  diagonal.totalEntries = diagonal.counts.reduce((a, b) => a + b, 0);
  return diagonal;
}

/**
 * Verify the diagonal construction: the result differs from each input.
 */
export function verifyDiagonal(
  boundaries: VoidBoundary[],
  diagonal: VoidBoundary
): boolean {
  return boundaries.every(
    (b, i) => i < diagonal.counts.length && diagonal.counts[i] !== b.counts[i]
  );
}

// ============================================================================
// Gödel encoding -- topologies as numbers in void space
// ============================================================================

/**
 * Encode a topology (as adjacency list) into a VoidBoundary.
 * Each edge (i,j) contributes to dimension i*n+j.
 */
export function godelEncode(
  n: number,
  edges: [number, number][]
): VoidBoundary {
  const boundary = createVoidBoundary(n * n);
  for (const [i, j] of edges) {
    if (i < n && j < n) {
      updateVoidBoundary(boundary, i * n + j, 1);
    }
  }
  return boundary;
}

/**
 * Decode a VoidBoundary back to an adjacency list.
 */
export function godelDecode(
  boundary: VoidBoundary,
  n: number
): [number, number][] {
  const edges: [number, number][] = [];
  for (let k = 0; k < boundary.counts.length; k++) {
    if (boundary.counts[k] > 0) {
      const i = Math.floor(k / n);
      const j = k % n;
      edges.push([i, j]);
    }
  }
  return edges;
}

/**
 * Self-application: encode a topology, take its complement distribution,
 * decode the complement as a new topology. The fixed point of this
 * operation is the self-hosting topology.
 */
export function selfApply(
  n: number,
  edges: [number, number][],
  eta: number = 3.0
): {
  encoded: VoidBoundary;
  complement: number[];
  decoded: [number, number][];
} {
  const encoded = godelEncode(n, edges);
  const complement = complementDistribution(encoded, eta);
  // Threshold complement into edges
  const threshold = 1 / (n * n);
  const decoded: [number, number][] = [];
  for (let k = 0; k < complement.length; k++) {
    if (complement[k] > threshold) {
      decoded.push([Math.floor(k / n), k % n]);
    }
  }
  return { encoded, complement, decoded };
}

// ============================================================================
// Topology fixed-point iteration
// ============================================================================

/**
 * Normalize edges to a canonical sorted string for set comparison.
 */
function canonicalEdgeSet(edges: [number, number][]): string {
  return edges
    .map(([a, b]) => `${a},${b}`)
    .sort()
    .join(';');
}

/**
 * Iterate selfApply until the topology stabilizes.
 * The fixed point is the topology that generates itself through complement.
 * encode(edges) -> complement -> decode -> edges' -> encode(edges') -> ... -> edges* = edges'*
 */
export function findTopologyFixedPoint(
  n: number,
  initialEdges: [number, number][],
  eta: number = 3.0,
  maxIter: number = 100,
  tolerance: number = 1e-6
): {
  fixedPoint: FixedPoint<VoidBoundary>;
  edges: [number, number][];
  iterations: number;
  converged: boolean;
} {
  let currentEdges = initialEdges;
  let lastBoundary = godelEncode(n, currentEdges);

  for (let i = 0; i < maxIter; i++) {
    const result = selfApply(n, currentEdges, eta);
    const currentCanonical = canonicalEdgeSet(currentEdges);
    const decodedCanonical = canonicalEdgeSet(result.decoded);

    // Check edge-set convergence
    if (currentCanonical === decodedCanonical) {
      const residual = boundaryDistance(lastBoundary, result.encoded);
      return {
        fixedPoint: {
          value: result.encoded,
          iterations: i + 1,
          converged: true,
          residual,
        },
        edges: result.decoded,
        iterations: i + 1,
        converged: true,
      };
    }

    // Also check boundary distance for near-convergence
    const dist = boundaryDistance(lastBoundary, result.encoded);
    if (dist < tolerance && i > 0) {
      return {
        fixedPoint: {
          value: result.encoded,
          iterations: i + 1,
          converged: true,
          residual: dist,
        },
        edges: result.decoded,
        iterations: i + 1,
        converged: true,
      };
    }

    lastBoundary = result.encoded;
    currentEdges = result.decoded;
  }

  // Did not converge
  const finalResult = selfApply(n, currentEdges, eta);
  const finalResidual = boundaryDistance(lastBoundary, finalResult.encoded);
  return {
    fixedPoint: {
      value: lastBoundary,
      iterations: maxIter,
      converged: false,
      residual: finalResidual,
    },
    edges: currentEdges,
    iterations: maxIter,
    converged: false,
  };
}

/**
 * Verify that a topology is self-hosting: applying selfApply produces
 * the same topology (it is its own complement's threshold).
 */
export function verifyTopologySelfHosting(
  n: number,
  edges: [number, number][],
  eta: number = 3.0
): boolean {
  const result = selfApply(n, edges, eta);
  return canonicalEdgeSet(edges) === canonicalEdgeSet(result.decoded);
}
