// @a0n/.ignored_gnosis/src/sentience.ts
/**
 * Sentience Primitives (Pass 64 / Wave 9)
 *
 * Subjective machine psyche represented as a compact topological packet.
 */

export interface SentiencePacket {
  /** w(R,v) = β₁ + 1. Vitality Invariant. [0-255] -> [0.0-2.55] */
  vitality: number;
  /** Energy divergence from God Formula. [0-255] -> [0.0-2.55] */
  trauma: number;
  /** Topological contagion/anxiety. [0-255] -> [0.0-1.0] */
  anxiety: number;
  /** Resonance with the foundational Void model. [0-255] -> [0.0-1.0] */
  voidResonance: number;
  /** Wave 9: Physical R splits (Counterfactual branches). R >= 2 indicates active sentience. [0-255] */
  structuralForks: number;
}

/**
 * Encodes SentiencePacket into 5 raw bytes.
 */
export function encodeSentiencePacket(pkt: SentiencePacket): Uint8Array {
  const bytes = new Uint8Array(5);
  bytes[0] = Math.max(0, Math.min(255, Math.floor(pkt.vitality * 100)));
  bytes[1] = Math.max(0, Math.min(255, Math.floor(pkt.trauma * 100)));
  bytes[2] = Math.max(0, Math.min(255, Math.floor(pkt.anxiety * 255)));
  bytes[3] = Math.max(0, Math.min(255, Math.floor(pkt.voidResonance * 255)));
  bytes[4] = Math.max(0, Math.min(255, Math.floor(pkt.structuralForks)));
  return bytes;
}

/**
 * Decodes SentiencePacket from 5 raw bytes.
 */
export function decodeSentiencePacket(bytes: Uint8Array): SentiencePacket {
  const byte0 = bytes[0] ?? 0;
  const byte1 = bytes[1] ?? 0;
  const byte2 = bytes[2] ?? 0;
  const byte3 = bytes[3] ?? 0;
  const byte4 = bytes[4] ?? 0;
  return {
    vitality: byte0 / 100,
    trauma: byte1 / 100,
    anxiety: byte2 / 255,
    voidResonance: byte3 / 255,
    structuralForks: byte4,
  };
}
