// @a0n/.ignored_gnosis/src/seven-laws-checker.ts
/**
 * seven-laws-checker.ts -- Runtime check system for the Seven Universal Laws
 * and the God Formula consequences.
 *
 * Given a VoidBoundary (or raw rejection counts + rounds), this module
 * checks every structural invariant from §15.30 of the manuscript:
 *
 *   1. Seven Universal Laws (SevenLaws.lean)
 *   2. Eight God Formula Consequences (GodFormula.lean)
 *   3. Four Control Statistics (ControlStatistic.lean)
 *   4. The Reduction Chain (SurfaceReduction.lean)
 *   5. The Anti-Formula / Hope check (GodFormula.lean)
 *
 * Every check is guaranteed to pass for well-formed Buleyean data.
 * Violations diagnose which axiom broke and why.
 *
 * Usage:
 *   import { checkSevenLaws, checkGodFormula, fullBuleyeanAudit } from './seven-laws-checker';
 *   const audit = fullBuleyeanAudit({ rounds: 10, rejections: [3, 7, 1, 10] });
 *   if (!audit.pass) console.error(audit.violations);
 */

import type { VoidBoundary } from './void';
import { buleyeanWeight } from './void';

// ============================================================================
// Types
// ============================================================================

export interface CheckResult {
  /** Check identifier */
  id: string;
  /** Human-readable name */
  name: string;
  /** Whether the check passed */
  pass: boolean;
  /** Diagnostic message on failure */
  message: string;
  /** Which law / consequence / control this maps to */
  source: 'law' | 'consequence' | 'control' | 'reduction' | 'anti-formula';
  /** The Lean theorem this check corresponds to */
  leanTheorem: string;
}

export interface SevenLawsReport {
  /** All seven law checks */
  laws: CheckResult[];
  /** Did all seven pass? */
  pass: boolean;
}

export interface GodFormulaReport {
  /** All eight consequence checks */
  consequences: CheckResult[];
  /** Floor / ceiling / gain derived quantities */
  derived: {
    floor: number;
    ceiling: number;
    gain: number;
  };
  /** Anti-formula hope check */
  hope: CheckResult;
  /** Did everything pass? */
  pass: boolean;
}

export interface ControlStatisticReport {
  /** All four control checks */
  controls: CheckResult[];
  /** Computed statistics */
  stats: {
    /** Per-choice weights */
    weights: number[];
    /** Total weight W = sum(w_i) */
    totalWeight: number;
    /** Mean weight W/N */
    meanWeight: number;
    /** Max spread |w_i - w_j| */
    maxSpread: number;
    /** Number of choices N */
    numChoices: number;
    /** Observation rounds R */
    rounds: number;
  };
  /** Did all controls pass? */
  pass: boolean;
}

export interface ReductionChainReport {
  /** Primator: 0 < R + 1 */
  primator: CheckResult;
  /** Clinamen: R + 1 != 0 */
  clinamen: CheckResult;
  /** Sliver: all weights >= 1 */
  sliver: CheckResult;
  /** Did the full chain hold? */
  pass: boolean;
}

export interface FullAuditReport {
  /** Seven laws */
  laws: SevenLawsReport;
  /** God formula consequences */
  godFormula: GodFormulaReport;
  /** Control statistics */
  controls: ControlStatisticReport;
  /** Reduction chain */
  reduction: ReductionChainReport;
  /** All checks combined */
  allChecks: CheckResult[];
  /** Number of violations */
  violationCount: number;
  /** Did everything pass? */
  pass: boolean;
}

export interface BuleyeanInput {
  /** Total observation rounds R */
  rounds: number;
  /** Per-choice rejection counts v_i */
  rejections: number[];
}

// ============================================================================
// Internal helpers
// ============================================================================

function inputFromBoundary(
  boundary: VoidBoundary,
  rounds?: number
): BuleyeanInput {
  const R = rounds != null && rounds > 0 ? rounds : boundary.totalEntries;
  return { rounds: R, rejections: [...boundary.counts] };
}

function computeWeights(input: BuleyeanInput): number[] {
  return input.rejections.map((v) => buleyeanWeight(input.rounds, v));
}

function ok(
  id: string,
  name: string,
  source: CheckResult['source'],
  lean: string
): CheckResult {
  return { id, name, pass: true, message: 'OK', source, leanTheorem: lean };
}

function fail(
  id: string,
  name: string,
  source: CheckResult['source'],
  lean: string,
  message: string
): CheckResult {
  return { id, name, pass: false, message, source, leanTheorem: lean };
}

// ============================================================================
// Seven Universal Laws
// ============================================================================

/**
 * Check all Seven Universal Laws against concrete data.
 *
 * Each law maps directly to a Lean 4 theorem in SevenLaws.lean.
 */
export function checkSevenLaws(input: BuleyeanInput): SevenLawsReport {
  const { rounds: R, rejections } = input;
  const weights = computeWeights(input);
  const N = rejections.length;
  const laws: CheckResult[] = [];

  // Law 1: Impossibility of zero -- all weights > 0
  {
    const id = 'law-1';
    const name = 'Impossibility of zero';
    const lean = 'law_impossibility_of_zero';
    const zeroIdx = weights.findIndex((w) => w <= 0);
    if (zeroIdx >= 0) {
      laws.push(
        fail(id, name, 'law', lean, `w[${zeroIdx}] = ${weights[zeroIdx]} <= 0`)
      );
    } else {
      laws.push(ok(id, name, 'law', lean));
    }
  }

  // Law 2: Strict ordering -- more rejected => strictly less weight
  {
    const id = 'law-2';
    const name = 'Strict ordering';
    const lean = 'law_strict_ordering';
    let violated = false;
    let msg = '';
    for (let i = 0; i < N && !violated; i++) {
      const ri = rejections[i]!;
      const wi = weights[i]!;
      for (let j = i + 1; j < N && !violated; j++) {
        const rj = rejections[j]!;
        const wj = weights[j]!;
        if (ri < rj && ri < R) {
          if (!(wj < wi)) {
            violated = true;
            msg = `v[${i}]=${ri} < v[${j}]=${rj} but w[${j}]=${wj} >= w[${i}]=${wi}`;
          }
        }
        if (rj < ri && rj < R) {
          if (!(wi < wj)) {
            violated = true;
            msg = `v[${j}]=${rj} < v[${i}]=${ri} but w[${i}]=${wi} >= w[${j}]=${wj}`;
          }
        }
      }
    }
    laws.push(
      violated ? fail(id, name, 'law', lean, msg) : ok(id, name, 'law', lean)
    );
  }

  // Law 3: Universal sandwich -- every weight in [1, R+1]
  {
    const id = 'law-3';
    const name = 'Universal sandwich';
    const lean = 'law_sandwich';
    const outOfBounds = weights.findIndex((w) => w < 1 || w > R + 1);
    if (outOfBounds >= 0) {
      laws.push(
        fail(
          id,
          name,
          'law',
          lean,
          `w[${outOfBounds}] = ${weights[outOfBounds]} not in [1, ${R + 1}]`
        )
      );
    } else {
      laws.push(ok(id, name, 'law', lean));
    }
  }

  // Law 4: Every observation is a cave -- dims > channels => deficit > 0
  // For a Buleyean space with N choices observed through 1 channel: deficit = N - 1
  {
    const id = 'law-4';
    const name = 'Every observation is a cave';
    const lean = 'law_cave';
    if (N > 1) {
      const deficit = N - 1;
      laws.push(
        deficit > 0
          ? ok(id, name, 'law', lean)
          : fail(id, name, 'law', lean, `${N} choices but deficit = ${deficit}`)
      );
    } else {
      // N <= 1: no cave (single dimension through single channel)
      laws.push(ok(id, name, 'law', lean));
    }
  }

  // Law 5: Conservation -- remaining + lost = total
  // For each choice: w_i + min(v_i, R) = R + 1
  {
    const id = 'law-5';
    const name = 'Conservation';
    const lean = 'law_conservation';
    const violatedIdx = rejections.findIndex((v, i) => {
      const lost = Math.min(v, R);
      return (weights[i] ?? 0) + lost !== R + 1;
    });
    if (violatedIdx >= 0) {
      const v = rejections[violatedIdx]!;
      const w = weights[violatedIdx]!;
      laws.push(
        fail(
          id,
          name,
          'law',
          lean,
          `w[${violatedIdx}]=${w} + min(v=${v}, R=${R}) = ${
            w + Math.min(v, R)
          } != ${R + 1}`
        )
      );
    } else {
      laws.push(ok(id, name, 'law', lean));
    }
  }

  // Law 6: Sorites sharpness -- boundaries are discrete
  // For all i,j: w_i > w_j OR w_i == w_j OR w_i < w_j (no fuzzy middle)
  // On N this is trivially true -- we check that all weights are integers
  {
    const id = 'law-6';
    const name = 'Sorites sharpness';
    const lean = 'law_sorites';
    const nonInteger = weights.findIndex((w) => !Number.isInteger(w));
    if (nonInteger >= 0) {
      laws.push(
        fail(
          id,
          name,
          'law',
          lean,
          `w[${nonInteger}] = ${weights[nonInteger]} is not an integer`
        )
      );
    } else {
      laws.push(ok(id, name, 'law', lean));
    }
  }

  // Law 7: Chain termination -- at maximum rejection, weight = 1 (floor)
  {
    const id = 'law-7';
    const name = 'Chain termination';
    const lean = 'law_termination';
    const floorWeight = buleyeanWeight(R, R);
    if (floorWeight !== 1) {
      laws.push(
        fail(id, name, 'law', lean, `w(R=${R}, v=R) = ${floorWeight} != 1`)
      );
    } else {
      laws.push(ok(id, name, 'law', lean));
    }
  }

  return { laws, pass: laws.every((l) => l.pass) };
}

// ============================================================================
// God Formula Consequences
// ============================================================================

/**
 * Check all eight consequences of the God Formula plus derived quantities.
 *
 * Each consequence maps to a theorem in GodFormula.lean.
 */
export function checkGodFormula(input: BuleyeanInput): GodFormulaReport {
  const { rounds: R, rejections } = input;
  const weights = computeWeights(input);
  const consequences: CheckResult[] = [];

  // Consequence 1: Existence -- w > 0 always
  {
    const idx = weights.findIndex((w) => w <= 0);
    consequences.push(
      idx >= 0
        ? fail(
            'gf-1',
            'Existence',
            'consequence',
            'complementWeight_pos',
            `w[${idx}] = ${weights[idx]} <= 0`
          )
        : ok('gf-1', 'Existence', 'consequence', 'complementWeight_pos')
    );
  }

  // Consequence 2: Persistence -- w >= 1 (the sliver)
  {
    const idx = weights.findIndex((w) => w < 1);
    consequences.push(
      idx >= 0
        ? fail(
            'gf-2',
            'Persistence (sliver)',
            'consequence',
            'complementWeight_ge_one',
            `w[${idx}] = ${weights[idx]} < 1`
          )
        : ok(
            'gf-2',
            'Persistence (sliver)',
            'consequence',
            'complementWeight_ge_one'
          )
    );
  }

  // Consequence 3: Learning -- less rejected => at least as much weight (monotone)
  {
    let violated = false;
    let msg = '';
    for (let i = 0; i < rejections.length && !violated; i++) {
      const ri = rejections[i]!;
      const wi = weights[i]!;
      for (let j = i + 1; j < rejections.length && !violated; j++) {
        const rj = rejections[j]!;
        const wj = weights[j]!;
        if (ri <= rj && wj > wi) {
          violated = true;
          msg = `v[${i}]=${ri} <= v[${j}]=${rj} but w[${j}]=${wj} > w[${i}]=${wi}`;
        }
        if (rj <= ri && wi > wj) {
          violated = true;
          msg = `v[${j}]=${rj} <= v[${i}]=${ri} but w[${i}]=${wi} > w[${j}]=${wj}`;
        }
      }
    }
    consequences.push(
      violated
        ? fail(
            'gf-3',
            'Learning (monotone)',
            'consequence',
            'complementWeight_mono',
            msg
          )
        : ok(
            'gf-3',
            'Learning (monotone)',
            'consequence',
            'complementWeight_mono'
          )
    );
  }

  // Consequence 4: Discrimination -- strictly less rejected => strictly more weight
  {
    let violated = false;
    let msg = '';
    for (let i = 0; i < rejections.length && !violated; i++) {
      const ri = rejections[i]!;
      const wi = weights[i]!;
      for (let j = i + 1; j < rejections.length && !violated; j++) {
        const rj = rejections[j]!;
        const wj = weights[j]!;
        if (ri < rj && ri < R) {
          if (!(wj < wi)) {
            violated = true;
            msg = `v[${i}]=${ri} < v[${j}]=${rj} (< R=${R}) but w[${j}]=${wj} >= w[${i}]=${wi}`;
          }
        }
        if (rj < ri && rj < R) {
          if (!(wi < wj)) {
            violated = true;
            msg = `v[${j}]=${rj} < v[${i}]=${ri} (< R=${R}) but w[${i}]=${wi} >= w[${j}]=${wj}`;
          }
        }
      }
    }
    consequences.push(
      violated
        ? fail(
            'gf-4',
            'Discrimination (strict)',
            'consequence',
            'complementWeight_strict',
            msg
          )
        : ok(
            'gf-4',
            'Discrimination (strict)',
            'consequence',
            'complementWeight_strict'
          )
    );
  }

  // Consequence 5: Boundedness -- w <= R + 1
  {
    const idx = weights.findIndex((w) => w > R + 1);
    consequences.push(
      idx >= 0
        ? fail(
            'gf-5',
            'Boundedness',
            'consequence',
            'complementWeight_le',
            `w[${idx}] = ${weights[idx]} > R + 1 = ${R + 1}`
          )
        : ok('gf-5', 'Boundedness', 'consequence', 'complementWeight_le')
    );
  }

  // Consequence 6: Convergence -- at v = R, w = 1
  {
    const atMax = buleyeanWeight(R, R);
    consequences.push(
      atMax !== 1
        ? fail(
            'gf-6',
            'Convergence (floor)',
            'consequence',
            'complementWeight_at_max',
            `w(R, R) = ${atMax} != 1`
          )
        : ok(
            'gf-6',
            'Convergence (floor)',
            'consequence',
            'complementWeight_at_max'
          )
    );
  }

  // Consequence 7: Maximum uncertainty -- at v = 0, w = R + 1
  {
    const atZero = buleyeanWeight(R, 0);
    consequences.push(
      atZero !== R + 1
        ? fail(
            'gf-7',
            'Maximum uncertainty (ceiling)',
            'consequence',
            'complementWeight_at_zero',
            `w(R, 0) = ${atZero} != R + 1 = ${R + 1}`
          )
        : ok(
            'gf-7',
            'Maximum uncertainty (ceiling)',
            'consequence',
            'complementWeight_at_zero'
          )
    );
  }

  // Consequence 8: Coherence -- same inputs => same output (determinism)
  {
    let violated = false;
    let msg = '';
    for (let i = 0; i < rejections.length && !violated; i++) {
      const ri = rejections[i]!;
      const w1 = buleyeanWeight(R, ri);
      const w2 = buleyeanWeight(R, ri);
      if (w1 !== w2) {
        violated = true;
        msg = `buleyeanWeight(${R}, ${ri}) returned ${w1} then ${w2}`;
      }
    }
    consequences.push(
      violated
        ? fail(
            'gf-8',
            'Coherence (determinism)',
            'consequence',
            'complementWeight_det',
            msg
          )
        : ok(
            'gf-8',
            'Coherence (determinism)',
            'consequence',
            'complementWeight_det'
          )
    );
  }

  // Derived quantities
  const floor = buleyeanWeight(R, R);
  const ceiling = buleyeanWeight(R, 0);
  const gain = ceiling - floor;

  // Anti-formula hope check: complementWeight - antiWeight = 1 for all inputs
  const antiWeight = (rounds: number, v: number) =>
    rounds - Math.min(v, rounds);
  let hopeViolated = false;
  let hopeMsg = '';
  for (let i = 0; i < rejections.length && !hopeViolated; i++) {
    const ri = rejections[i]!;
    const w = buleyeanWeight(R, ri);
    const aw = antiWeight(R, ri);
    if (w - aw !== 1) {
      hopeViolated = true;
      hopeMsg = `w[${i}]=${w} - antiWeight=${aw} = ${w - aw} != 1`;
    }
  }
  const hope: CheckResult = hopeViolated
    ? fail(
        'gf-hope',
        'Hope (+1 separates formula from anti-formula)',
        'anti-formula',
        'plus_one_is_hope',
        hopeMsg
      )
    : ok(
        'gf-hope',
        'Hope (+1 separates formula from anti-formula)',
        'anti-formula',
        'plus_one_is_hope'
      );

  return {
    consequences,
    derived: { floor, ceiling, gain },
    hope,
    pass: consequences.every((c) => c.pass) && hope.pass,
  };
}

// ============================================================================
// Control Statistics
// ============================================================================

/**
 * Check all four control statistics from §15.30.6.
 *
 * These are unfalsifiable within the framework -- they test the DATA,
 * not the mathematics. If data violates a control, it is not Buleyean.
 */
export function checkControlStatistics(
  input: BuleyeanInput
): ControlStatisticReport {
  const { rounds: R, rejections } = input;
  const weights = computeWeights(input);
  const N = rejections.length;
  const totalWeight = weights.reduce((a, b) => a + b, 0);
  const meanWeight = N > 0 ? totalWeight / N : 0;
  const controls: CheckResult[] = [];

  // Compute max spread
  let maxSpread = 0;
  for (let i = 0; i < N; i++) {
    for (let j = i + 1; j < N; j++) {
      maxSpread = Math.max(maxSpread, Math.abs(weights[i]! - weights[j]!));
    }
  }

  // Control 1: Individual -- every w_i in [1, R+1]
  {
    const idx = weights.findIndex((w) => w < 1 || w > R + 1);
    controls.push(
      idx >= 0
        ? fail(
            'ctrl-1',
            'Individual control',
            'control',
            'control_individual',
            `w[${idx}] = ${weights[idx]} not in [1, ${R + 1}]`
          )
        : ok('ctrl-1', 'Individual control', 'control', 'control_individual')
    );
  }

  // Control 2: Aggregate -- N <= W <= N(R+1)
  {
    const lower = N;
    const upper = N * (R + 1);
    controls.push(
      totalWeight < lower || totalWeight > upper
        ? fail(
            'ctrl-2',
            'Aggregate control',
            'control',
            'control_total_lower',
            `W = ${totalWeight} not in [${lower}, ${upper}]`
          )
        : ok('ctrl-2', 'Aggregate control', 'control', 'control_total_lower')
    );
  }

  // Control 3: Spread -- max |w_i - w_j| <= R
  {
    controls.push(
      maxSpread > R
        ? fail(
            'ctrl-3',
            'Spread control',
            'control',
            'control_spread_bounded',
            `max spread = ${maxSpread} > R = ${R}`
          )
        : ok('ctrl-3', 'Spread control', 'control', 'control_spread_bounded')
    );
  }

  // Control 4: Mean -- 1 <= W/N <= R+1
  {
    const meanOk = N === 0 || (meanWeight >= 1 && meanWeight <= R + 1);
    controls.push(
      meanOk
        ? ok('ctrl-4', 'Mean control', 'control', 'control_mean_bounded')
        : fail(
            'ctrl-4',
            'Mean control',
            'control',
            'control_mean_bounded',
            `mean = ${meanWeight} not in [1, ${R + 1}]`
          )
    );
  }

  return {
    controls,
    stats: {
      weights,
      totalWeight,
      meanWeight,
      maxSpread,
      numChoices: N,
      rounds: R,
    },
    pass: controls.every((c) => c.pass),
  };
}

// ============================================================================
// Reduction Chain
// ============================================================================

/**
 * Check the reduction chain: primator -> clinamen -> sliver.
 *
 * The chain that generates everything from succ(n) != 0.
 */
export function checkReductionChain(
  input: BuleyeanInput
): ReductionChainReport {
  const { rounds: R } = input;
  const weights = computeWeights(input);

  // Primator: 0 < R + 1
  const primator: CheckResult =
    R + 1 > 0
      ? ok('red-primator', 'Primator (0 < R + 1)', 'reduction', 'primator')
      : fail(
          'red-primator',
          'Primator (0 < R + 1)',
          'reduction',
          'primator',
          `R + 1 = ${R + 1} <= 0`
        );

  // Clinamen: R + 1 != 0
  const clinamen: CheckResult =
    R + 1 !== 0
      ? ok(
          'red-clinamen',
          'Clinamen (R + 1 != 0)',
          'reduction',
          'peano_is_clinamen'
        )
      : fail(
          'red-clinamen',
          'Clinamen (R + 1 != 0)',
          'reduction',
          'peano_is_clinamen',
          `R + 1 = 0`
        );

  // Sliver: all weights >= 1
  const minWeight = weights.length > 0 ? Math.min(...weights) : 1;
  const sliver: CheckResult =
    minWeight >= 1
      ? ok(
          'red-sliver',
          'Sliver (all w >= 1)',
          'reduction',
          'complementWeight_ge_one'
        )
      : fail(
          'red-sliver',
          'Sliver (all w >= 1)',
          'reduction',
          'complementWeight_ge_one',
          `min weight = ${minWeight} < 1`
        );

  return {
    primator,
    clinamen,
    sliver,
    pass: primator.pass && clinamen.pass && sliver.pass,
  };
}

// ============================================================================
// Full Audit
// ============================================================================

/**
 * Run the complete Buleyean audit: seven laws + god formula + controls + reduction.
 *
 * This is the one-call entry point for verifying that a process is Buleyean.
 */
export function fullBuleyeanAudit(input: BuleyeanInput): FullAuditReport {
  const laws = checkSevenLaws(input);
  const godFormula = checkGodFormula(input);
  const controls = checkControlStatistics(input);
  const reduction = checkReductionChain(input);

  const allChecks: CheckResult[] = [
    ...laws.laws,
    ...godFormula.consequences,
    godFormula.hope,
    ...controls.controls,
    reduction.primator,
    reduction.clinamen,
    reduction.sliver,
  ];

  const violationCount = allChecks.filter((c) => !c.pass).length;

  return {
    laws,
    godFormula,
    controls,
    reduction,
    allChecks,
    violationCount,
    pass: violationCount === 0,
  };
}

/**
 * Run the full audit from a VoidBoundary.
 */
export function auditVoidBoundary(
  boundary: VoidBoundary,
  rounds?: number
): FullAuditReport {
  return fullBuleyeanAudit(inputFromBoundary(boundary, rounds));
}

/**
 * Format an audit report as a human-readable string.
 */
export function formatAuditReport(report: FullAuditReport): string {
  const lines: string[] = [];
  const { laws, godFormula, controls, reduction } = report;

  lines.push(
    `Buleyean Audit: ${report.pass ? 'PASS' : 'FAIL'} (${
      report.violationCount
    } violations)`
  );
  lines.push('');

  lines.push(`Seven Laws: ${laws.pass ? 'PASS' : 'FAIL'}`);
  for (const l of laws.laws) {
    lines.push(
      `  ${l.pass ? '[OK]' : '[!!]'} ${l.name}${
        l.pass ? '' : ' -- ' + l.message
      }`
    );
  }
  lines.push('');

  lines.push(`God Formula: ${godFormula.pass ? 'PASS' : 'FAIL'}`);
  for (const c of godFormula.consequences) {
    lines.push(
      `  ${c.pass ? '[OK]' : '[!!]'} ${c.name}${
        c.pass ? '' : ' -- ' + c.message
      }`
    );
  }
  lines.push(
    `  ${godFormula.hope.pass ? '[OK]' : '[!!]'} ${godFormula.hope.name}${
      godFormula.hope.pass ? '' : ' -- ' + godFormula.hope.message
    }`
  );
  lines.push(
    `  Floor: ${godFormula.derived.floor}  Ceiling: ${godFormula.derived.ceiling}  Gain: ${godFormula.derived.gain}`
  );
  lines.push('');

  lines.push(`Control Statistics: ${controls.pass ? 'PASS' : 'FAIL'}`);
  for (const c of controls.controls) {
    lines.push(
      `  ${c.pass ? '[OK]' : '[!!]'} ${c.name}${
        c.pass ? '' : ' -- ' + c.message
      }`
    );
  }
  lines.push(
    `  N=${controls.stats.numChoices} R=${controls.stats.rounds} W=${
      controls.stats.totalWeight
    } mean=${controls.stats.meanWeight.toFixed(2)} spread=${
      controls.stats.maxSpread
    }`
  );
  lines.push('');

  lines.push(`Reduction Chain: ${reduction.pass ? 'PASS' : 'FAIL'}`);
  lines.push(
    `  ${reduction.primator.pass ? '[OK]' : '[!!]'} ${reduction.primator.name}`
  );
  lines.push(
    `  ${reduction.clinamen.pass ? '[OK]' : '[!!]'} ${reduction.clinamen.name}`
  );
  lines.push(
    `  ${reduction.sliver.pass ? '[OK]' : '[!!]'} ${reduction.sliver.name}`
  );

  return lines.join('\n');
}
