// @a0n/.ignored_gnosis/src/slowness-certificate.ts
/**
 * Slowness Certificate — Leg 3 of the Temporal Prism trilogy.
 *
 * A SlownessCertificate proves that an operation took AT LEAST minLatencyMs
 * of wall-clock time. This is the temporal dual of a proof-of-work: instead
 * of proving you did N units of compute, you prove you waited N units of time.
 *
 * The proof construct is the attestation chain:
 *   - Every hop is signed by a named PID (Ed25519)
 *   - Timestamps are monotonic (enforced by chain verification)
 *   - The sum of inter-hop elapsed time is the certified lower bound
 *   - The issuer wraps the chain with a signature over the whole certificate
 *
 * Use cases:
 *   - Time-locked commitments (arrive no sooner than X)
 *   - Anti-flash-loan / anti-MEV (transactions must age before execution)
 *   - Witness-aged votes (proposals must circulate for a minimum period)
 *   - Rate-limit proofs (prove you waited between retries)
 *   - Auditable latency SLAs (prove a held payload was actually held)
 *
 * The cert is transport-agnostic. What's provable is the *temporal lower
 * bound*, not the mechanism used to achieve it.
 *
 * @packageDocumentation
 */

import type {
  AttestationSigner,
  AttestationVerifier,
  ChainBuildContext,
  PrismAttestationChain,
  AttestationChainVerification,
} from './prism-attestation';
import { verifyAttestationChain } from './prism-attestation';

// ─── Types ──────────────────────────────────────────────────────────

/**
 * A SlownessCertificate proves an operation took at least minLatencyMs.
 *
 * Soundness rests on four independent checks:
 *   1. The attestation chain is valid (signatures + monotonic timestamps)
 *   2. The chain's total elapsed time >= minLatencyMs
 *   3. The issuer signed the certificate bytes
 *   4. The issuer's DID is trusted by the verifier (out of band)
 */
export interface SlownessCertificate {
  readonly source: 'GnosisSlownessCertificate';
  readonly certificateId: string;
  /** Hash of the operation being certified. */
  readonly operationHash: string;
  /** The proven lower bound on elapsed time, in milliseconds. */
  readonly minLatencyMs: number;
  /** The observed total elapsed time across the chain, in milliseconds. */
  readonly actualLatencyMs: number;
  /** The underlying attestation chain that witnesses the elapsed time. */
  readonly attestationChain: PrismAttestationChain;
  /** DID of the PID that issued (sealed) the certificate. */
  readonly issuerDid: string;
  /** Picolorenzo timestamp at issuance. */
  readonly issuedAtPLo: number;
  /** Ed25519 signature over canonicalCertificateBytes(cert). */
  readonly issuerSignature: Uint8Array;
}

/**
 * Structured verification result for a slowness certificate.
 */
export interface SlownessCertificateVerification {
  readonly source: 'GnosisSlownessCertificateVerification';
  readonly certificateId: string;
  readonly valid: boolean;
  /** The claimed minLatencyMs is met by the chain's total elapsed time. */
  readonly minLatencyMet: boolean;
  /** The underlying attestation chain passes verification. */
  readonly chainValid: boolean;
  /** The issuer's signature over the cert bytes is valid. */
  readonly issuerSignatureValid: boolean;
  /** The actualLatencyMs field matches the chain's observed elapsed time. */
  readonly actualLatencyConsistent: boolean;
  readonly reasons: readonly string[];
  readonly chainVerification?: AttestationChainVerification;
}

export interface SlownessCertificateVerifier {
  verify(cert: SlownessCertificate): Promise<SlownessCertificateVerification>;
}

// ─── Canonical Serialization ────────────────────────────────────────

/**
 * Canonical byte encoding of the certificate's signed body.
 *
 * Format: UTF-8 of
 *   "certificateId|operationHash|minLatencyMs|actualLatencyMs|chainId|issuerDid|issuedAtPLo|finalPayloadHash|hopCount"
 *
 * This is what the issuer signs, and what the verifier re-hashes to check
 * the issuer signature. Every field that binds the cert to its chain and
 * its claim is included. A tampered field produces a different canonical
 * byte string and breaks signature verification.
 *
 * Note: the attestation chain's own signatures are NOT re-signed by the
 * issuer; the chain is self-certifying. The issuer signs a *summary* that
 * pins the chain identity (chainId + finalPayloadHash + hopCount) plus the
 * claimed bound.
 */
export function canonicalCertificateBytes(cert: SlownessCertificate): Uint8Array {
  const parts = [
    cert.certificateId,
    cert.operationHash,
    cert.minLatencyMs.toString(),
    cert.actualLatencyMs.toString(),
    cert.attestationChain.chainId,
    cert.issuerDid,
    cert.issuedAtPLo.toString(),
    cert.attestationChain.finalPayloadHash,
    cert.attestationChain.hops.length.toString(),
  ];
  return new TextEncoder().encode(parts.join('|'));
}

// ─── Chain Elapsed Time ─────────────────────────────────────────────

/**
 * Compute the elapsed time witnessed by an attestation chain.
 *
 * Elapsed = lastHop.attestedAtPLo - firstHop.attestedAtPLo.
 *
 * Timestamps are in picolorenzos, which we treat as milliseconds for the
 * purposes of this certificate (1 pLo = 1 ms at the default clock). If
 * your deployment uses a different unit, scale the inputs accordingly.
 *
 * Returns 0 for empty chains or chains with a single hop (no interval).
 */
export function chainElapsedMs(chain: PrismAttestationChain): number {
  if (chain.hops.length < 2) return 0;
  const first = chain.hops[0].attestedAtPLo;
  const last = chain.hops[chain.hops.length - 1].attestedAtPLo;
  return Math.max(0, last - first);
}

// ─── Issuance ───────────────────────────────────────────────────────

export interface IssueCertificateInput {
  readonly certificateId: string;
  readonly operationHash: string;
  readonly minLatencyMs: number;
  readonly attestationChain: PrismAttestationChain;
  readonly issuer: AttestationSigner;
  readonly issuedAtPLo: number;
}

/**
 * Issue a SlownessCertificate over a completed attestation chain.
 *
 * The caller has built and verified the chain already (see
 * `prism-attestation.ts`). The issuer reads the elapsed time from the
 * chain, stamps the claimed lower bound, and signs the canonical bytes.
 *
 * Throws if the chain does not witness enough elapsed time to support the
 * claim. Issuers MUST refuse to sign bounds they cannot prove; producing
 * an unsound cert is the only way this primitive fails silently.
 */
export async function issueSlownessCertificate(
  input: IssueCertificateInput
): Promise<SlownessCertificate> {
  const { certificateId, operationHash, minLatencyMs, attestationChain,
          issuer, issuedAtPLo } = input;

  if (minLatencyMs < 0) {
    throw new Error(`minLatencyMs must be non-negative; got ${minLatencyMs}`);
  }
  if (attestationChain.hops.length < 2) {
    throw new Error('Attestation chain must have at least 2 hops to witness elapsed time');
  }

  const actualLatencyMs = chainElapsedMs(attestationChain);
  if (actualLatencyMs < minLatencyMs) {
    throw new Error(
      `Chain elapsed time ${actualLatencyMs}ms is less than claimed minLatencyMs ${minLatencyMs}ms; refusing to issue unsound certificate`
    );
  }

  // Build the pre-signature cert so we can canonicalize it.
  const preSignatureCert: SlownessCertificate = {
    source: 'GnosisSlownessCertificate',
    certificateId,
    operationHash,
    minLatencyMs,
    actualLatencyMs,
    attestationChain,
    issuerDid: issuer.did,
    issuedAtPLo,
    issuerSignature: new Uint8Array(0), // placeholder, signed below
  };

  const canonical = canonicalCertificateBytes(preSignatureCert);
  const issuerSignature = await issuer.sign(canonical);

  return {
    ...preSignatureCert,
    issuerSignature,
  };
}

// ─── Verification ───────────────────────────────────────────────────

/**
 * Verify a SlownessCertificate.
 *
 * Four independent gates, reported individually so callers can debug
 * which component failed:
 *   1. chainValid: the attestation chain passes verifyAttestationChain
 *   2. minLatencyMet: chainElapsedMs(chain) >= cert.minLatencyMs
 *   3. issuerSignatureValid: the issuer's signature over the cert bytes
 *      checks out
 *   4. actualLatencyConsistent: cert.actualLatencyMs matches
 *      chainElapsedMs(chain) (prevents lying about the observed value)
 *
 * Overall `valid` is the conjunction. Any single failure invalidates the
 * certificate.
 */
export async function verifySlownessCertificate(
  cert: SlownessCertificate,
  chainContext: ChainBuildContext,
  issuerVerifier: AttestationVerifier
): Promise<SlownessCertificateVerification> {
  const reasons: string[] = [];

  // 1. Chain validation.
  const chainVerification = await verifyAttestationChain(cert.attestationChain, chainContext);
  const chainValid = chainVerification.valid;
  if (!chainValid) {
    reasons.push(`Attestation chain invalid: ${chainVerification.reasons.join('; ')}`);
  }

  // 2. Min latency met.
  const observedElapsedMs = chainElapsedMs(cert.attestationChain);
  const minLatencyMet = observedElapsedMs >= cert.minLatencyMs;
  if (!minLatencyMet) {
    reasons.push(
      `Chain elapsed ${observedElapsedMs}ms does not meet claimed minLatencyMs ${cert.minLatencyMs}ms`
    );
  }

  // 3. actualLatencyMs matches what the chain actually witnesses.
  const actualLatencyConsistent = cert.actualLatencyMs === observedElapsedMs;
  if (!actualLatencyConsistent) {
    reasons.push(
      `actualLatencyMs ${cert.actualLatencyMs}ms does not match observed ${observedElapsedMs}ms`
    );
  }

  // 4. Issuer signature.
  const canonical = canonicalCertificateBytes(cert);
  let issuerSignatureValid = false;
  try {
    issuerSignatureValid = await issuerVerifier.verify(
      cert.issuerDid, canonical, cert.issuerSignature
    );
  } catch (err) {
    reasons.push(`Issuer verifier threw: ${String(err)}`);
  }
  if (!issuerSignatureValid) {
    reasons.push(`Issuer signature invalid for DID ${cert.issuerDid}`);
  }

  const valid = chainValid && minLatencyMet && actualLatencyConsistent && issuerSignatureValid;

  return {
    source: 'GnosisSlownessCertificateVerification',
    certificateId: cert.certificateId,
    valid,
    minLatencyMet,
    chainValid,
    issuerSignatureValid,
    actualLatencyConsistent,
    reasons,
    chainVerification,
  };
}
