// @a0n/.ignored_gnosis/src/solar-geometry.ts
/**
 * Solar Geometry — re-exports core solar position functions from math.ts
 * plus the Kepler-precision ecliptic path for aeon-solarium integration.
 *
 * The core solar geometry lives in math.ts as part of the Gnostic Time System.
 * This module re-exports those functions and adds the ecliptic-position path
 * that accepts Earth's heliocentric vector from Kepler propagation.
 *
 * @packageDocumentation
 */

// Re-export core solar geometry from the Gnostic Time System
export type {
  TwilightZone,
  SolarObserver,
  SolarPosition,
} from './math';

export {
  OBLIQUITY_DEG,
  solarDeclination,
  equationOfTime,
  classifyTwilightZone,
  twilightZoneFraction,
  solarPosition,
  isDaylightAtPLo,
  solarPositionAtPLo,
} from './math';

// ─── Types ──────────────────────────────────────────────────────────

/** 3D vector (matches aeon-solarium's Vector3 contract). */
export interface Vector3 {
  readonly x: number;
  readonly y: number;
  readonly z: number;
}

// For backward compat with the topology module
export type SolarAltitude = import('./math.js').SolarPosition;

// ─── Constants ──────────────────────────────────────────────────────

const DEG_TO_RAD = Math.PI / 180;
const RAD_TO_DEG = 180 / Math.PI;
const OBLIQUITY_RAD = 23.4393 * DEG_TO_RAD;
const J2000_EPOCH_MS = Date.UTC(2000, 0, 1, 12, 0, 0);
const MS_PER_DAY = 86_400_000;

import { classifyTwilightZone as _classify, twilightZoneFraction as _fraction } from './math';
import type { SolarObserver as _Observer, SolarPosition as _Position } from './math';

/** Normalize degrees to [0, 360). */
function normalizeDeg(deg: number): number {
  return ((deg % 360) + 360) % 360;
}

// ─── Solar Altitude (Ecliptic / Kepler) ─────────────────────────────

/**
 * Compute solar altitude from Earth's heliocentric ecliptic position.
 *
 * Accepts Earth's position (in AU, heliocentric ecliptic frame) as returned
 * by aeon-solarium's `propagateSolarSystemOrbit(earthOrbit, atIso)`.
 * The sun's geocentric position is the negation of Earth's heliocentric vector.
 *
 * Higher precision than the analytic day-of-year approximation because
 * it captures orbital eccentricity through the Kepler solver.
 */
export function solarPositionFromEcliptic(
  earthHeliocentricAu: Vector3,
  observer: _Observer,
  epochMs: number
): _Position {
  // Sun's geocentric ecliptic position
  const sunX = -earthHeliocentricAu.x;
  const sunZ = -earthHeliocentricAu.z;
  const sunY = -earthHeliocentricAu.y;

  // Ecliptic longitude and latitude of the sun
  const eclLonRad = Math.atan2(sunZ, sunX);
  const eclLatRad = Math.atan2(sunY, Math.sqrt(sunX * sunX + sunZ * sunZ));

  // Ecliptic to equatorial: declination
  const sinDec =
    Math.sin(eclLatRad) * Math.cos(OBLIQUITY_RAD) +
    Math.cos(eclLatRad) * Math.sin(OBLIQUITY_RAD) * Math.sin(eclLonRad);
  const decRad = Math.asin(Math.max(-1, Math.min(1, sinDec)));

  // Right ascension
  const raRad = Math.atan2(
    Math.sin(eclLonRad) * Math.cos(OBLIQUITY_RAD) -
      Math.tan(eclLatRad) * Math.sin(OBLIQUITY_RAD),
    Math.cos(eclLonRad)
  );

  // Greenwich Mean Sidereal Time (degrees)
  const julianDays = (epochMs - J2000_EPOCH_MS) / MS_PER_DAY;
  const gmstDeg = normalizeDeg(280.46061837 + 360.98564736629 * julianDays);
  const lstDeg = normalizeDeg(gmstDeg + observer.longitudeDeg);

  // Hour angle
  const haRad = (lstDeg - raRad * RAD_TO_DEG) * DEG_TO_RAD;

  // Altitude
  const latRad = observer.latitudeDeg * DEG_TO_RAD;
  const sinAlt =
    Math.sin(latRad) * Math.sin(decRad) +
    Math.cos(latRad) * Math.cos(decRad) * Math.cos(haRad);
  const altitudeDeg = Math.asin(Math.max(-1, Math.min(1, sinAlt))) * RAD_TO_DEG;

  // Azimuth
  const cosAlt = Math.cos(altitudeDeg * DEG_TO_RAD);
  const cosAz = cosAlt > 0.001
    ? (Math.sin(decRad) - Math.sin(latRad) * sinAlt) / (Math.cos(latRad) * cosAlt)
    : 0;
  let azimuthDeg = Math.acos(Math.max(-1, Math.min(1, cosAz))) * RAD_TO_DEG;
  if (Math.sin(haRad) > 0) azimuthDeg = 360 - azimuthDeg;

  const zone = _classify(altitudeDeg);
  return {
    altitudeDeg,
    azimuthDeg: normalizeDeg(azimuthDeg),
    zone,
    zoneFraction: _fraction(altitudeDeg),
  };
}

// Alias for backward compatibility with daylight-topology imports
export { solarPosition as solarAltitude } from './math';
