// @a0n/.ignored_gnosis/src/spiderweb-topology.ts
/**
 * Spiderweb Mesh Network — Orb Web Topology as Mesh Architecture
 *
 * Maps the spiderweb topology (β₁ = k·s) directly to mesh networking:
 *   - Radials (k) = peer connections / streams
 *   - Spiral rings (s) = relay hop depth / pipeline stages
 *   - Hub = coordinator / gateway node
 *   - β₁ = k·s = independent routing paths
 *   - Δβ = routing deficit in Bules
 *
 * Gnostic mesh configurations prescribe architectures at each scale,
 * from Bythos (isolated node) through Pleroma (55-path full mesh).
 *
 * @see ../lean/Lean/ForkRaceFoldTheorems/SpiderwebMeshNetwork.lean for formal proofs
 * @packageDocumentation
 */

// ─── Types ──────────────────────────────────────────────────────────

export interface MeshConfig {
  /** Number of direct peer connections (radials) */
  readonly peers: number;
  /** Relay hop depth (spiral rings) */
  readonly hops: number;
}

export type GnosticMeshLevel =
  | 'bythos'
  | 'barbelo'
  | 'syzygy'
  | 'proton'
  | 'primitives'
  | 'emanations'
  | 'sophia'
  | 'kenoma'
  | 'void'
  | 'pleroma';

export type MeshRegime = 'laminar' | 'transitional' | 'turbulent';

export interface MeshTopologyReport {
  readonly config: MeshConfig;
  readonly level: GnosticMeshLevel | null;
  readonly routingPaths: number;
  readonly wireCount: number;
  readonly nodeCount: number;
  readonly regime: MeshRegime;
  readonly reynoldsEstimate: number;
  readonly peerFailureCost: number;
  readonly relayFailureCost: number;
}

export interface MeshDeficitReport {
  readonly required: number;
  readonly available: number;
  readonly deficit: number;
  readonly utilization: number;
  readonly tcpWaste: number;
  readonly assessment: string;
}

export interface MeshUpgradePath {
  readonly from: GnosticMeshLevel;
  readonly to: GnosticMeshLevel;
  readonly currentPaths: number;
  readonly targetPaths: number;
  readonly deficit: number;
  readonly additionalWires: number;
  readonly additionalNodes: number;
}

/**
 * Projection frame for shape-vs-unit accounting.
 * Mirrors Lean surface: ProjectionParadoxUnits.lean
 */
export interface ProjectionFrame {
  /** Observed chart dimension (post-projection). */
  readonly observedDim: number;
  /** Latent/evolved chart dimension (pre-projection). */
  readonly latentDim: number;
}

/**
 * Minimal witness bundle for projection paradox detection.
 * Mirrors Lean definition of ProjectionParadox witnesses.
 */
export interface ProjectionParadoxWitness {
  /** Whether any observed state appears solved after projection. */
  readonly observedHasSolution: boolean;
  /** Whether a solved latent state exists before projection. */
  readonly latentHasSolution: boolean;
  /** Whether solved/unsolved latent states collapse to one observation. */
  readonly hasCollapsedSplit: boolean;
}

export interface ProjectionShapeUnitReport {
  readonly frame: ProjectionFrame;
  readonly dimensionalUnitGap: number;
  readonly hasDimensionalUnitGap: boolean;
  readonly reducedObservedLatentRatio: readonly [number, number];
  readonly isProjectionParadox: boolean;
  /** Atomic paradox units (OneParadox witnesses). */
  readonly paradoxUnits: number;
  /** Structural shape witness shorthand (units > 0). */
  readonly hasShape: boolean;
}

export interface ParadoxUnitSandwichReport {
  /** Predictive lower bound from frame alone. */
  readonly predictedLowerBound: number;
  /** Observed paradox units from witness evidence. */
  readonly observedUnits: number;
  /** Dimensional upper bound from latent-observed gap. */
  readonly dimensionalUpperBound: number;
  /** Whether lower <= observed <= upper. */
  readonly withinBounds: boolean;
}

export interface ProjectionPredictiveReport {
  readonly frame: ProjectionFrame;
  readonly predictedLowerBound: number;
  readonly observedUnits: number;
  /** observed - predicted */
  readonly residual: number;
  /** Exact predictive hit (residual == 0). */
  readonly consistent: boolean;
}

export interface ReductionProjectionInput {
  readonly id: string;
  readonly frame: ProjectionFrame;
  readonly witness: ProjectionParadoxWitness;
}

export interface ReductionProjectionScanReport {
  readonly id: string;
  readonly shapeUnit: ProjectionShapeUnitReport;
  readonly sandwich: ParadoxUnitSandwichReport;
  readonly predictive: ProjectionPredictiveReport;
}

// ─── Gnostic Mesh Configurations ────────────────────────────────────

export const GNOSTIC_MESHES: Record<GnosticMeshLevel, MeshConfig> = {
  bythos: { peers: 0, hops: 0 },
  barbelo: { peers: 1, hops: 1 },
  syzygy: { peers: 2, hops: 1 },
  proton: { peers: 3, hops: 1 },
  primitives: { peers: 5, hops: 1 },
  emanations: { peers: 3, hops: 2 },
  sophia: { peers: 3, hops: 3 },
  kenoma: { peers: 5, hops: 2 },
  void: { peers: 7, hops: 3 },
  pleroma: { peers: 11, hops: 5 },
};

export const GNOSTIC_BETA1: Record<GnosticMeshLevel, number> = {
  bythos: 0,
  barbelo: 1,
  syzygy: 2,
  proton: 3,
  primitives: 5,
  emanations: 6,
  sophia: 9,
  kenoma: 10,
  void: 21,
  pleroma: 55,
};

const GNOSTIC_ORDER: GnosticMeshLevel[] = [
  'bythos',
  'barbelo',
  'syzygy',
  'proton',
  'primitives',
  'emanations',
  'sophia',
  'kenoma',
  'void',
  'pleroma',
];

/**
 * Phyle cell composition for each gnostic level.
 * compositeDimension = (cells - 1) * 9 + 1.
 * sophia=1 cell (dim 1), kenoma=2 (dim 10), void=3 (dim 19), pleroma=7 (dim 55 exact).
 */
export const PHYLE_COMPOSITIONS: Record<GnosticMeshLevel, number> = {
  bythos: 0,
  barbelo: 1,
  syzygy: 1,
  proton: 1,
  primitives: 1,
  emanations: 1,
  sophia: 1,
  kenoma: 2,
  void: 3,
  pleroma: 7,
};

// ─── Core Computations ──────────────────────────────────────────────

/** β₁ = k·s: independent routing paths in the mesh. */
export function meshRoutingPaths(m: MeshConfig): number {
  return m.peers * m.hops;
}

/** Edge count: total wire-equivalents needed. E = k·(2s + 1). */
export function meshWireCount(m: MeshConfig): number {
  return m.peers * (2 * m.hops + 1);
}

/** Node count: coordinator + relays + endpoints. V = k·s + k + 1. */
export function meshNodeCount(m: MeshConfig): number {
  return m.peers * m.hops + m.peers + 1;
}

/** Reynolds estimate: k/s. Classifies mesh regime. */
export function meshReynolds(m: MeshConfig): number {
  if (m.hops === 0) return m.peers === 0 ? 0 : Infinity;
  return m.peers / m.hops;
}

/** Classify mesh regime from Reynolds estimate. */
export function classifyMeshRegime(m: MeshConfig): MeshRegime {
  const re = meshReynolds(m);
  if (re < 0.3) return 'laminar';
  if (re > 0.7) return 'turbulent';
  return 'transitional';
}

/** Routing deficit: required − available, floored at 0. */
export function meshDeficit(required: number, m: MeshConfig): number {
  return Math.max(0, required - meshRoutingPaths(m));
}

/** Cost of losing one peer (radial severance) = hops. */
export function peerFailureCost(m: MeshConfig): number {
  return m.hops;
}

/** Cost of losing one relay depth (spiral break) = peers. */
export function relayFailureCost(m: MeshConfig): number {
  return m.peers;
}

function clampNat(value: number): number {
  if (!Number.isFinite(value)) return 0;
  return Math.max(0, Math.trunc(value));
}

function gcdNat(a: number, b: number): number {
  let x = clampNat(a);
  let y = clampNat(b);
  while (y !== 0) {
    const t = x % y;
    x = y;
    y = t;
  }
  return x;
}

/** Hidden dimensional units between observed and latent frames. */
export function dimensionalUnitGap(frame: ProjectionFrame): number {
  const observed = clampNat(frame.observedDim);
  const latent = clampNat(frame.latentDim);
  return Math.max(0, latent - observed);
}

/** Whether the frame has any hidden dimensional unit. */
export function hasDimensionalUnitGap(frame: ProjectionFrame): boolean {
  const observed = clampNat(frame.observedDim);
  const latent = clampNat(frame.latentDim);
  return observed < latent;
}

/** Reduced observed:latent ratio (e.g., 2:54 -> 1:27). */
export function reducedObservedLatentRatio(
  frame: ProjectionFrame
): readonly [number, number] {
  const observed = clampNat(frame.observedDim);
  const latent = clampNat(frame.latentDim);
  const g = gcdNat(observed, latent);
  if (g === 0) return [0, 0] as const;
  return [Math.trunc(observed / g), Math.trunc(latent / g)] as const;
}

/** Projection paradox detector from explicit witness flags. */
export function isProjectionParadox(witness: ProjectionParadoxWitness): boolean {
  return !witness.observedHasSolution && witness.latentHasSolution && witness.hasCollapsedSplit;
}

/** Atomic paradox units: one witnessed projection paradox contributes one unit. */
export function projectionParadoxUnits(witness: ProjectionParadoxWitness): number {
  return isProjectionParadox(witness) ? 1 : 0;
}

/** Predictive lower bound from frame only (no witness): positive gap -> 1, else 0. */
export function predictParadoxUnitLowerBoundFromFrame(frame: ProjectionFrame): number {
  return hasDimensionalUnitGap(frame) ? 1 : 0;
}

/** Sandwich theorem runtime mirror: predictedLowerBound <= observedUnits <= dimensionalUpperBound. */
export function paradoxUnitSandwich(
  frame: ProjectionFrame,
  witness: ProjectionParadoxWitness
): ParadoxUnitSandwichReport {
  const predictedLowerBound = predictParadoxUnitLowerBoundFromFrame(frame);
  const observedUnits = projectionParadoxUnits(witness);
  const dimensionalUpperBound = dimensionalUnitGap(frame);
  const withinBounds =
    predictedLowerBound <= observedUnits && observedUnits <= dimensionalUpperBound;
  return {
    predictedLowerBound,
    observedUnits,
    dimensionalUpperBound,
    withinBounds,
  };
}

/** Predictive diagnostic for the frame model against observed witness units. */
export function analyzeProjectionPredictive(
  frame: ProjectionFrame,
  witness: ProjectionParadoxWitness
): ProjectionPredictiveReport {
  const predictedLowerBound = predictParadoxUnitLowerBoundFromFrame(frame);
  const observedUnits = projectionParadoxUnits(witness);
  const residual = observedUnits - predictedLowerBound;
  return {
    frame,
    predictedLowerBound,
    observedUnits,
    residual,
    consistent: residual === 0,
  };
}

/**
 * Cross-reduction scan:
 * evaluate shape/unit, sandwich bounds, and predictive residual across many reductions.
 */
export function scanReductionProjections(
  reductions: readonly ReductionProjectionInput[]
): ReductionProjectionScanReport[] {
  return reductions.map((reduction) => {
    const shapeUnit = analyzeProjectionShapeUnit(reduction.frame, reduction.witness);
    const sandwich = paradoxUnitSandwich(reduction.frame, reduction.witness);
    const predictive = analyzeProjectionPredictive(reduction.frame, reduction.witness);
    return {
      id: reduction.id,
      shapeUnit,
      sandwich,
      predictive,
    };
  });
}

/**
 * Full shape-vs-unit report:
 * - shape is topological (collapsed solved/unsolved split),
 * - unit is cardinal (atomic paradox unit count),
 * - dimensional gap tracks latent lift budget.
 */
export function analyzeProjectionShapeUnit(
  frame: ProjectionFrame,
  witness: ProjectionParadoxWitness
): ProjectionShapeUnitReport {
  const paradoxUnits = projectionParadoxUnits(witness);
  return {
    frame,
    dimensionalUnitGap: dimensionalUnitGap(frame),
    hasDimensionalUnitGap: hasDimensionalUnitGap(frame),
    reducedObservedLatentRatio: reducedObservedLatentRatio(frame),
    isProjectionParadox: paradoxUnits > 0,
    paradoxUnits,
    hasShape: paradoxUnits > 0,
  };
}

// ─── Predictive Dimensional Metric ────────────────────────────────────

/** Lower Bound: Floor conditions for minimal operation.
 *  A mesh is degenerate or sequential if it lacks multi-peer structure.
 */
export function requiresSpiderwebLowerBound(concurrency: number, hops: number): boolean {
  return concurrency <= 1 || (concurrency === 2 && hops === 0);
}

/** Gain Threshold: When spherical geometric routing saturation begins.
 *  Requires at least 2 independent radials to form a meaningful fork/fold.
 */
export function yieldsSpiderwebGainThreshold(concurrency: number, hops: number): boolean {
  return concurrency >= 2 && hops >= 1;
}

/** Ceiling: Torus loop recurrence / fullness limit.
 *  Saturation of the physical routing constraints (Pleroma level).
 */
export function requiresSpiderwebCeiling(concurrency: number, hops: number): boolean {
  return concurrency >= 11 && hops >= 5;
}

/**
 * Predicts the optimal GnosticLevel scheduling dimension (Spiderweb topology)
 * based on mathematical structural workloads (concurrency, hops, BFT needs).
 * @param concurrency Map to k (peers)
 * @param hops Map to s (cycles/feedback)
 * @param bftFaults Estimated faults (f) to tolerate
 * @param saturationReached Is the maximum available bandwidth/topology saturated
 */
export function predictSpiderwebDimension(
  concurrency: number,
  hops: number,
  bftFaults: number,
  saturationReached: boolean = false
): GnosticMeshLevel {
  if (saturationReached || requiresSpiderwebCeiling(concurrency, hops)) {
    return 'pleroma';
  } else if (bftFaults >= 2 || concurrency >= 7 || (concurrency === 5 && hops >= 3)) {
    return 'void';
  } else if (bftFaults === 1 && hops >= 2) {
    return 'kenoma';
  } else if (hops >= 3 && concurrency <= 3) {
    return 'sophia';
  } else if (hops === 2 && concurrency === 3) {
    return 'emanations';
  } else if (concurrency >= 5 && hops === 1) {
    return 'primitives';
  } else if (concurrency >= 3 && hops === 1) {
    return 'proton';
  } else if (yieldsSpiderwebGainThreshold(concurrency, hops)) {
    return 'syzygy';
  } else if (concurrency === 1 && hops === 1) {
    return 'barbelo';
  } else {
    return 'bythos';
  }
}

/**
 * Automagically resolves the appropriate Syzygy dimensional scale based on
 * the number of parameters. (10 dimensions supported: 0D up to 55D).
 * < 3D falls back to basic gates (useful for bits or simple conditionals).
 * 
 * @param numParameters Number of parameters in the workload (e.g. 70e9)
 */
export function resolveSyzygyDimension(numParameters: number): GnosticMeshLevel {
  if (numParameters >= 50_000_000_000) return 'pleroma'; // 55D
  if (numParameters >= 10_000_000_000) return 'void';    // 21D
  if (numParameters >= 3_000_000_000) return 'kenoma';   // 10D
  if (numParameters >= 1_000_000_000) return 'sophia';   // 9D
  if (numParameters >= 500_000_000) return 'emanations'; // 6D
  if (numParameters >= 100_000_000) return 'primitives'; // 5D
  if (numParameters >= 10_000_000) return 'proton';      // 3D
  if (numParameters >= 1_000_000) return 'syzygy';       // 2D
  if (numParameters >= 100_000) return 'barbelo';        // 1D
  return 'bythos';                                       // 0D (for bits or degenerate logic)
}

/**
 * Generates the GG topology source for a Dynamic Syzygy Cannon.
 * It automatically dilates to the specified dimensionality.
 * 
 * @param dimension The GnosticMeshLevel to scale the cannon to.
 */
export function generateDynamicSyzygyCannonTopology(dimension: GnosticMeshLevel): string {
  const beta1 = GNOSTIC_BETA1[dimension];
  
  if (beta1 <= 0) {
    // 0D degenerate bypass (Bythos)
    return [
      `// Syzygy Cannon (Dimension 0D - Bythos)`,
      `(source:Source)-[:PROCESS]->(sink:Sink)`
    ].join('\n');
  }

  const lines: string[] = [];
  lines.push(`// Syzygy Cannon (Dimension ${beta1}D - ${dimension})`);
  lines.push(`// Automagically dilated to appropriate dimensionality.`);
  
  // Feed layer workers
  const feeds = Array.from({ length: beta1 }, (_, i) => `F${i}`);
  const feedDecl = feeds.map(f => `${f}:CannonFeed`).join(' | ');
  
  // Reap layer workers
  const reaps = Array.from({ length: beta1 }, (_, i) => `R${i}`);
  
  // 1. Source forks to Feed
  lines.push(`(source:Source)-[:FORK]->(${feedDecl})`);
  
  // 2. Feed pairs with Reap (the Cannon spiral alignment)
  for (let i = 0; i < beta1; i++) {
    lines.push(`(F${i})-[:PROCESS]->(R${i})`);
  }
  
  // 3. Reap folds to sink
  lines.push(`(${reaps.join(' | ')})-[:FOLD { strategy: 'syzygy-align' }]->(sink:Sink)`);
  
  return lines.join('\n');
}

// ─── Composition — Meta-Scheduler Topology ─────────────────────────

export type CompositionLayout =
  | 'ring' | 'chain' | 'star' | 'ladder' | 'helix'
  | 'grid' | 'complete' | 'tree';

export interface CompositionDecomposition {
  /** Base scheduler gnostic level */
  readonly baseLevel: GnosticMeshLevel;
  /** Base dimension (β₁) */
  readonly baseDim: number;
  /** Number of schedulers needed */
  readonly count: number;
  /** Composite dimension: (count-1) * baseDim + 1 */
  readonly compositeDim: number;
  /** Whether the composition lands exactly on the target */
  readonly exact: boolean;
}

export interface ComposedTopology {
  /** The decomposition used */
  readonly decomposition: CompositionDecomposition;
  /** Recommended layout for the web count */
  readonly layout: CompositionLayout;
  /** Recommended layer count for V-NAND stacking */
  readonly layers: number;
  /** Recommended curvature (0=flat, 1=sphere) */
  readonly curvature: number;
  /** Total β₁ across all schedulers */
  readonly totalBeta1: number;
  /** GG topology source */
  readonly topology: string;
}

/**
 * Composite dimension: (n-1)·d + 1.
 * Universal for ALL connected bonding topologies (proven in SchedulerComposition.lean).
 */
export function compositeDimension(schedulerCount: number, baseDim: number): number {
  if (schedulerCount <= 0) return 0;
  return (schedulerCount - 1) * baseDim + 1;
}

/**
 * Decompose a target dimension into n copies of a base dimension.
 * Returns n = (target - 1) / baseDim + 1, or null if it doesn't divide evenly.
 */
export function decomposeIntoCopies(targetDim: number, baseDim: number): number | null {
  if (baseDim <= 0 || targetDim <= 0) return null;
  const remainder = (targetDim - 1) % baseDim;
  if (remainder !== 0) return null;
  return (targetDim - 1) / baseDim + 1;
}

/**
 * Find all valid decompositions of a target dimension into gnostic base levels.
 * Returns them sorted by efficiency (fewest total schedulers first).
 */
export function findDecompositions(targetDim: number): CompositionDecomposition[] {
  const results: CompositionDecomposition[] = [];

  for (const level of GNOSTIC_ORDER) {
    const baseDim = GNOSTIC_BETA1[level];
    if (baseDim <= 0) continue;

    const count = decomposeIntoCopies(targetDim, baseDim);
    if (count !== null && count >= 2) {
      results.push({
        baseLevel: level,
        baseDim,
        count,
        compositeDim: targetDim,
        exact: true,
      });
    }
  }

  // Sort by count (fewest schedulers = most efficient slingshot)
  results.sort((a, b) => a.count - b.count);
  return results;
}

/**
 * Find the optimal decomposition for a parameter count.
 * Uses resolveSyzygyDimension to get the target, then finds the
 * most efficient composition path to reach it.
 *
 * The slingshot effect: intermediate dimensions amplify reach,
 * so composing through a mid-level gnostic number uses fewer
 * total schedulers than going direct from Barbelo.
 */
export function optimalComposition(numParameters: number): ComposedTopology {
  const targetLevel = resolveSyzygyDimension(numParameters);
  const targetDim = GNOSTIC_BETA1[targetLevel];

  // Find all exact decompositions
  const decomps = findDecompositions(targetDim);

  // Pick the best: fewest copies, but prefer base dimensions >= 3 (BFT-capable)
  const bftDecomps = decomps.filter(d => d.baseDim >= 3);
  const best = (bftDecomps.length > 0 ? bftDecomps[0] : decomps[0]) ?? {
    baseLevel: targetLevel,
    baseDim: targetDim,
    count: 1,
    compositeDim: targetDim,
    exact: true,
  };

  // Choose layout based on web count
  let layout: CompositionLayout;
  if (best.count <= 2) layout = 'chain';
  else if (best.count === 3) layout = 'ring';       // Triangle — wallington rotation
  else if (best.count <= 5) layout = 'ring';         // Pentagon self-tile
  else if (best.count <= 7) layout = 'ring';
  else layout = 'grid';

  // Layer recommendation: if count > 7, stack into layers
  const websPerLayer = Math.min(best.count, 7);
  const layers = Math.max(1, Math.ceil(best.count / websPerLayer));

  // Curvature: higher dimensions → more spherical
  const curvature = Math.min(1, best.baseDim / 10);

  // Generate GG topology
  const topologyLines: string[] = [];
  topologyLines.push(`// Composed Meta-Scheduler: ${best.count}× ${best.baseLevel} (${best.baseDim}D) → ${best.compositeDim}D`);
  topologyLines.push(`// Layout: ${layout}, Layers: ${layers}, Curvature: ${curvature.toFixed(2)}`);
  topologyLines.push(`// Universal composition law: (${best.count}-1)·${best.baseDim} + 1 = ${best.compositeDim}`);

  // Source forks to web coordinators
  const webNames = Array.from({ length: websPerLayer }, (_, i) => `W${i}`);
  topologyLines.push(`(source:Source)-[:FORK]->(${webNames.map(w => `${w}:WebCoordinator`).join(' | ')})`);

  // Each web is a full spiderweb mesh at the base dimension
  const baseMesh = GNOSTIC_MESHES[best.baseLevel];
  for (const w of webNames) {
    topologyLines.push(`(${w})-[:MESH { peers: ${baseMesh.peers}, hops: ${baseMesh.hops}, beta1: ${best.baseDim} }]->(${w}_out:WebOutput)`);
  }

  // Outputs fold to sink with composition
  topologyLines.push(`(${webNames.map(w => `${w}_out`).join(' | ')})-[:FOLD { strategy: 'composition', compositeDim: ${best.compositeDim} }]->(sink:Sink)`);

  if (layers > 1) {
    topologyLines.push(`// V-NAND: ${layers} layers, pole-to-pole vias between layers`);
    topologyLines.push(`// Total schedulers: ${best.count}, Total β₁: ${best.baseDim * best.count}`);
  }

  return {
    decomposition: best,
    layout,
    layers,
    curvature,
    totalBeta1: best.baseDim * best.count,
    topology: topologyLines.join('\n'),
  };
}

/**
 * Self-composition: d copies of d-dimensional scheduler → d²-d+1.
 * The hella-vortex — the scheduler folds in on itself.
 */
export function selfCompose(baseDim: number): number {
  return compositeDimension(baseDim, baseDim);
}

/**
 * Slingshot path: find the minimum-step ascent from d=0 to targetDim
 * using composition steps with at most maxCopies per step.
 */
export function slingshotPath(targetDim: number, maxCopies: number = 7): number[] {
  const path: number[] = [0];
  let current = 0;

  while (current < targetDim) {
    // Try each copy count from maxCopies down to 2
    let jumped = false;
    for (let n = maxCopies; n >= 2; n--) {
      const next = compositeDimension(n, current);
      if (next <= targetDim) {
        path.push(next);
        current = next;
        jumped = true;
        break;
      }
    }
    if (!jumped) {
      // Can't reach exactly; take smallest step
      const next = compositeDimension(2, current);
      path.push(next);
      current = next;
    }
    // Safety: prevent infinite loop
    if (path.length > 100) break;
  }

  return path;
}

// ─── Analysis ───────────────────────────────────────────────────────

/** Identify which gnostic level a mesh config matches, if any. */
export function identifyGnosticLevel(m: MeshConfig): GnosticMeshLevel | null {
  for (const level of GNOSTIC_ORDER) {
    const gm = GNOSTIC_MESHES[level];
    if (gm.peers === m.peers && gm.hops === m.hops) return level;
  }
  return null;
}

/** Find the smallest gnostic mesh that covers a given routing demand. */
export function smallestAdequateMesh(requiredPaths: number): GnosticMeshLevel {
  for (const level of GNOSTIC_ORDER) {
    if (GNOSTIC_BETA1[level] >= requiredPaths) return level;
  }
  return 'pleroma';
}

/** Full topology report for a mesh configuration. */
export function analyzeMesh(m: MeshConfig): MeshTopologyReport {
  return {
    config: m,
    level: identifyGnosticLevel(m),
    routingPaths: meshRoutingPaths(m),
    wireCount: meshWireCount(m),
    nodeCount: meshNodeCount(m),
    regime: classifyMeshRegime(m),
    reynoldsEstimate: meshReynolds(m),
    peerFailureCost: peerFailureCost(m),
    relayFailureCost: relayFailureCost(m),
  };
}

/** Deficit report for a mesh against a routing demand. */
export function analyzeMeshDeficit(
  requiredPaths: number,
  m: MeshConfig
): MeshDeficitReport {
  const available = meshRoutingPaths(m);
  const deficit = Math.max(0, requiredPaths - available);
  const utilization =
    requiredPaths === 0 ? 1 : Math.min(1, available / requiredPaths);
  const tcpWaste = Math.max(0, requiredPaths - 1);

  let assessment: string;
  if (deficit === 0) {
    assessment = `Topology-matched: 0 Bules — mesh provides ${available} paths for ${requiredPaths} required`;
  } else if (utilization >= 0.8) {
    assessment = `Near-matched: ${deficit} Bules — ${(utilization * 100).toFixed(0)}% coverage`;
  } else if (available === 0) {
    assessment = `No mesh: ${deficit} Bules — isolated node cannot route`;
  } else {
    assessment = `Underprovisioned: ${deficit} Bules — mesh provides ${available} of ${requiredPaths} required paths`;
  }

  return { required: requiredPaths, available, deficit, utilization, tcpWaste, assessment };
}

/** Compute the upgrade path between two gnostic mesh levels. */
export function meshUpgradePath(
  from: GnosticMeshLevel,
  to: GnosticMeshLevel
): MeshUpgradePath {
  const fromConfig = GNOSTIC_MESHES[from];
  const toConfig = GNOSTIC_MESHES[to];
  return {
    from,
    to,
    currentPaths: meshRoutingPaths(fromConfig),
    targetPaths: meshRoutingPaths(toConfig),
    deficit: meshDeficit(meshRoutingPaths(toConfig), fromConfig),
    additionalWires: meshWireCount(toConfig) - meshWireCount(fromConfig),
    additionalNodes: meshNodeCount(toConfig) - meshNodeCount(fromConfig),
  };
}

/** Generate the complete gnostic mesh ladder with all reports. */
export function gnosticMeshLadder(): Map<GnosticMeshLevel, MeshTopologyReport> {
  const ladder = new Map<GnosticMeshLevel, MeshTopologyReport>();
  for (const level of GNOSTIC_ORDER) {
    ladder.set(level, analyzeMesh(GNOSTIC_MESHES[level]));
  }
  return ladder;
}
