// @a0n/.ignored_gnosis/src/structured-primitives.ts
import { GNOSTIC_MESHES } from './spiderweb-topology';
import { resolveWallingtonDimension } from './wallington-rotation';

type StructuredPrimitiveKind =
  | 'StructuredMoA'
  | 'WallingtonRotation'
  | 'WorthingtonWhip'
  | 'HeteroMoAFabric';

interface StructuredPrimitive {
  readonly id: string;
  readonly kind: StructuredPrimitiveKind;
  readonly properties: Record<string, string>;
}

const STRUCTURED_PRIMITIVE_PATTERN =
  /^\(\s*([A-Za-z_][A-Za-z0-9_]*)\s*:\s*(StructuredMoA|WallingtonRotation|WorthingtonWhip|HeteroMoAFabric)\s*(?:{([^}]*)})?\s*\)\s*$/;
const EDGE_PATTERN =
  /^\(([^)]+)\)\s*-\[:([A-Z]+)(?:\s*{([^}]*)})?\]->\s*\(([^)]+)\)\s*$/;
const ARRAY_LITERAL_PATTERN = /^\[(.*)\]$/;
const MAX_STRUCTURED_PRIMITIVE_PASSES = 8;

function countOccurrences(input: string, token: string): number {
  return input.split(token).length - 1;
}

function collectStructuredPrimitiveUnits(source: string): string[] {
  const lines = source.split('\n');
  const units: string[] = [];

  for (let index = 0; index < lines.length; index++) {
    const line = lines[index] ?? '';
    const trimmed = line.trim();

    if (
      trimmed.length === 0 ||
      trimmed.includes('-[:') ||
      !/\(\s*[A-Za-z_][A-Za-z0-9_]*\s*:\s*(StructuredMoA|WallingtonRotation|WorthingtonWhip|HeteroMoAFabric)\b/.test(
        trimmed
      )
    ) {
      units.push(line);
      continue;
    }

    let balance = countOccurrences(line, '(') - countOccurrences(line, ')');
    let endIndex = index;
    const buffer = [trimmed];

    while (balance > 0 && endIndex + 1 < lines.length) {
      endIndex += 1;
      const nextLine = lines[endIndex] ?? '';
      buffer.push(nextLine.trim());
      balance +=
        countOccurrences(nextLine, '(') - countOccurrences(nextLine, ')');
    }

    const candidate = buffer.join(' ');
    if (parseStructuredPrimitive(candidate)) {
      units.push(candidate);
      index = endIndex;
      continue;
    }

    units.push(line);
  }

  return units;
}

function readAliasedProperty(
  properties: Readonly<Record<string, string>>,
  ...keys: readonly string[]
): string | undefined {
  for (const key of keys) {
    const value = properties[key];
    if (typeof value === 'string' && value.length > 0) {
      return value;
    }
  }

  return undefined;
}

function parseProperties(propertiesRaw?: string): Record<string, string> {
  if (!propertiesRaw) {
    return {};
  }

  const properties: Record<string, string> = {};
  const pairs = propertiesRaw.match(
    /(\w+)\s*:\s*('[^']*'|"[^"]*"|\[[^\]]*\]|[^,]+)/g
  );
  if (!pairs) {
    return properties;
  }

  for (const pair of pairs) {
    const separator = pair.indexOf(':');
    if (separator < 0) {
      continue;
    }

    const key = pair.slice(0, separator).trim();
    const value = pair
      .slice(separator + 1)
      .trim()
      .replace(/^['"]|['"]$/g, '');

    if (key.length > 0 && value.length > 0) {
      properties[key] = value;
    }
  }

  return properties;
}

function parsePositiveInt(
  rawValue: string | undefined,
  fallback: number
): number {
  const parsed = Number(rawValue);
  if (!Number.isFinite(parsed) || parsed <= 0) {
    return fallback;
  }

  return Math.max(1, Math.floor(parsed));
}

function parseNonnegativeInt(
  rawValue: string | undefined,
  fallback: number
): number {
  const parsed = Number(rawValue);
  if (!Number.isFinite(parsed) || parsed < 0) {
    return fallback;
  }

  return Math.max(0, Math.floor(parsed));
}

function parseNameList(rawValue: string | undefined): readonly string[] {
  if (!rawValue) {
    return [];
  }

  const match = rawValue.match(ARRAY_LITERAL_PATTERN);
  if (!match) {
    return [];
  }

  return (match[1] ?? '')
    .split(',')
    .map((entry) => entry.trim().replace(/^['"]|['"]$/g, ''))
    .filter((entry) => entry.length > 0);
}

function resolveNameList(
  rawValue: string | undefined,
  count: number,
  fallbackName: (index: number) => string
): readonly string[] {
  const names = parseNameList(rawValue);
  if (names.length === count) {
    return names;
  }

  return Array.from({ length: count }, (_, index) => fallbackName(index));
}

function parseStructuredPrimitive(line: string): StructuredPrimitive | null {
  const match = line.match(STRUCTURED_PRIMITIVE_PATTERN);
  if (!match) {
    return null;
  }

  const id = match[1]?.trim();
  const kind = match[2]?.trim() as StructuredPrimitiveKind | undefined;
  if (!id || !kind) {
    return null;
  }

  return {
    id,
    kind,
    properties: parseProperties(match[3]?.trim()),
  };
}

function normalizeTargetSegment(
  segment: string,
  primitiveIds: ReadonlySet<string>
): string {
  const trimmed = segment.trim();
  const idMatch = trimmed.match(/^([A-Za-z_][A-Za-z0-9_]*)/);
  const id = idMatch?.[1];
  if (!id || !primitiveIds.has(id)) {
    return segment;
  }

  const replacement = `${id}__ingress`;
  return segment.replace(id, replacement);
}

function rewriteIncomingPrimitiveTargets(
  line: string,
  primitiveIds: ReadonlySet<string>
): string {
  const match = line.match(EDGE_PATTERN);
  if (!match) {
    return line;
  }

  const sourceRaw = match[1]?.trim() ?? '';
  const edgeType = match[2]?.trim() ?? '';
  const propertiesRaw = match[3]?.trim();
  const targetRaw = match[4]?.trim() ?? '';
  const rewrittenTargets = targetRaw
    .split('|')
    .map((segment) => normalizeTargetSegment(segment, primitiveIds))
    .join(' | ');

  const renderedProperties =
    propertiesRaw && propertiesRaw.length > 0 ? ` { ${propertiesRaw} }` : '';

  return `(${sourceRaw})-[:${edgeType}${renderedProperties}]->(${rewrittenTargets})`;
}

function renderPrimitiveProperties(
  properties: Readonly<Record<string, string>>
): string {
  const entries = Object.entries(properties);
  if (entries.length === 0) {
    return '';
  }

  return ` { ${entries
    .map(([key, value]) => `${key}: '${value}'`)
    .join(', ')} }`;
}

function renderEdgeProperties(
  properties: Readonly<Record<string, string>>
): string {
  return renderPrimitiveProperties(properties);
}

function renderNode(
  id: string,
  label: string,
  properties: Readonly<Record<string, string>>
): string {
  return `(${id}: ${label}${renderPrimitiveProperties(properties)})`;
}

function expandStructuredMoaPrimitive(
  primitive: StructuredPrimitive
): readonly string[] {
  const blocks = parsePositiveInt(primitive.properties.blocks, 4);
  const activeBlocks = Math.min(
    blocks,
    parsePositiveInt(
      primitive.properties.activeBlocks ?? primitive.properties.active_blocks,
      2
    )
  );
  const heads = parsePositiveInt(primitive.properties.heads, 4);
  const activeHeads = Math.min(
    heads,
    parsePositiveInt(
      primitive.properties.activeHeads ?? primitive.properties.active_heads,
      2
    )
  );
  const stages = parsePositiveInt(primitive.properties.stages, 4);
  const chunks = parsePositiveInt(primitive.properties.chunks, 2);
  const blockParameters = String(
    parsePositiveInt(primitive.properties.parameters, 18)
  );
  const headParameters = String(
    parsePositiveInt(primitive.properties.headParameters, 4)
  );
  const blockLabel = primitive.properties.blockLabel ?? 'MoATransformerlet';
  const headLabel = primitive.properties.headLabel ?? 'AttentionHeadChain';
  const outputLabel = primitive.properties.outputLabel ?? 'Tensor';
  const family = primitive.properties.family ?? 'moa';
  const innerFoldStrategy =
    primitive.properties.innerFoldStrategy ??
    primitive.properties.headFoldStrategy ??
    'linear';
  const outerFoldStrategy =
    primitive.properties.outerFoldStrategy ??
    primitive.properties.strategy ??
    'linear';
  const corridorBase =
    readAliasedProperty(primitive.properties, 'corridor', 'corridor_base') ??
    primitive.id;
  const corridorMode =
    readAliasedProperty(
      primitive.properties,
      'corridorMode',
      'corridor_mode'
    ) ?? 'readwrite';
  const reuseScope =
    readAliasedProperty(primitive.properties, 'reuseScope', 'reuse_scope') ??
    'corridor';
  const failureMode = primitive.properties.failure ?? 'vent';
  const traceLabel =
    readAliasedProperty(primitive.properties, 'traceLabel', 'trace_label') ??
    'CorridorTrace';
  const ventLabel =
    readAliasedProperty(primitive.properties, 'ventLabel', 'vent_label') ??
    'VoidBoundary';
  const ventCondition =
    readAliasedProperty(
      primitive.properties,
      'ventCondition',
      'vent_condition'
    ) ?? 'corridor-reject';
  const blockNames = resolveNameList(
    primitive.properties.blockNames,
    blocks,
    (index) => `${primitive.id}__block_${index}`
  );
  const outputProperties = renderPrimitiveProperties({
    primitive: 'StructuredMoA',
    blocks: String(blocks),
    active_blocks: String(activeBlocks),
    heads: String(heads),
    active_heads: String(activeHeads),
    family,
  });

  const lines: string[] = [];
  lines.push(
    `// StructuredMoA '${primitive.id}' expanded into rotated concurrent MoA topology with MiddleOut request compression corridors.`
  );
  const outerWhip = `${primitive.id}__outer_whip`;
  const outerTrace = `${primitive.id}__outer_trace`;
  const outerVoid = `${primitive.id}__outer_void`;
  const outerCorridor = `${corridorBase}/outer-whip`;
  lines.push(
    `(${primitive.id}__ingress: FlowFrame { role: 'structured-moa-ingress', primitive: 'StructuredMoA' })`
  );
  lines.push(
    `(${primitive.id}__outer_rotation: RotationScheduler { schedule: 'wallington_rotation', stages: '${stages}', chunks: '${chunks}', role: 'outer-transformer-rotation' })`
  );
  lines.push(
    `(${primitive.id}__router: RoutingGate { transformerlets: '${blocks}', active_paths: '${activeBlocks}', role: 'outer-moa-router' })`
  );
  lines.push(
    renderNode(outerWhip, 'FoldOperator', {
      role: 'worthington-whip',
      boundary: 'outer-whip',
      strategy: outerFoldStrategy,
    })
  );
  lines.push(
    renderNode(outerTrace, traceLabel, {
      role: 'middle-out-request-compression-trace',
      corridor: outerCorridor,
      boundary: 'outer-whip',
      primitive: 'StructuredMoA',
    })
  );
  lines.push(
    renderNode(outerVoid, ventLabel, {
      role: 'middle-out-request-compression-void',
      corridor: outerCorridor,
      boundary: 'outer-whip',
      primitive: 'StructuredMoA',
    })
  );
  lines.push(`(${primitive.id}: ${outputLabel}${outputProperties})`);
  lines.push(
    `(${primitive.id}__ingress)-[:PROCESS]->(${primitive.id}__outer_rotation)`
  );
  lines.push(
    `(${primitive.id}__outer_rotation)-[:FORK { schedule: 'wallington_rotation' }]->(${primitive.id}__router)`
  );
  lines.push(`(${primitive.id}__router)-[:FORK]->(${blockNames.join(' | ')})`);

  const blockOutputs: string[] = [];
  for (const [blockIndex, blockName] of blockNames.entries()) {
    const headRotation = `${blockName}__head_rotation`;
    const headRouter = `${blockName}__head_router`;
    const headWhip = `${blockName}__head_whip`;
    const headTrace = `${blockName}__head_trace`;
    const headVoid = `${blockName}__head_void`;
    const blockOut = `${blockName}__out`;
    const innerCorridor = `${corridorBase}/${blockName}/head-whip`;
    const headNames = Array.from(
      { length: heads },
      (_, headIndex) => `${blockName}__h${headIndex}`
    );
    blockOutputs.push(blockOut);

    lines.push(
      `(${blockName}: ${blockLabel} { parameters: '${blockParameters}', heads: '${heads}', active_heads: '${activeHeads}', family: '${family}', block_index: '${blockIndex}' })`
    );
    lines.push(
      `(${headRotation}: RotationScheduler { schedule: 'wallington_rotation', stages: '${stages}', chunks: '${chunks}', role: 'inner-head-rotation' })`
    );
    lines.push(
      `(${headRouter}: HeadRouter { heads: '${heads}', active_heads: '${activeHeads}', block: '${blockName}' })`
    );
    for (const [headIndex, headName] of headNames.entries()) {
      lines.push(
        `(${headName}: ${headLabel} { parameters: '${headParameters}', head: '${headIndex}', block: '${blockName}' })`
      );
    }
    lines.push(
      `(${headWhip}: FoldOperator { role: 'worthington-whip', boundary: 'head-whip', strategy: '${innerFoldStrategy}' })`
    );
    lines.push(
      renderNode(headTrace, traceLabel, {
        role: 'middle-out-request-compression-trace',
        corridor: innerCorridor,
        boundary: 'head-whip',
        block: blockName,
        primitive: 'StructuredMoA',
      })
    );
    lines.push(
      renderNode(headVoid, ventLabel, {
        role: 'middle-out-request-compression-void',
        corridor: innerCorridor,
        boundary: 'head-whip',
        block: blockName,
        primitive: 'StructuredMoA',
      })
    );
    lines.push(`(${blockOut}: FlowFrame { role: 'transformerlet-out' })`);
    lines.push(`(${blockName})-[:PROCESS]->(${headRotation})`);
    lines.push(
      `(${headRotation})-[:FORK { schedule: 'wallington_rotation' }]->(${headRouter})`
    );
    lines.push(`(${headRouter})-[:FORK]->(${headNames.join(' | ')})`);
    lines.push(
      `(${headNames.join(' | ')})-[:FOLD${renderEdgeProperties({
        strategy: innerFoldStrategy,
        boundary: 'head-whip',
        corridor: innerCorridor,
        corridor_mode: corridorMode,
        reuse_scope: reuseScope,
        failure: failureMode,
      })}]->(${headWhip})`
    );
    lines.push(`(${headWhip})-[:PROCESS]->(${headTrace})`);
    lines.push(
      `(${headWhip})-[:VENT${renderEdgeProperties({
        condition: ventCondition,
        corridor: innerCorridor,
        repair_debt: '0',
        drift_coefficient: '0',
      })}]->(${headVoid})`
    );
    lines.push(`(${headTrace})-[:PROCESS]->(${blockOut})`);
  }

  lines.push(
    `(${blockOutputs.join(' | ')})-[:FOLD${renderEdgeProperties({
      strategy: outerFoldStrategy,
      boundary: 'outer-whip',
      corridor: outerCorridor,
      corridor_mode: corridorMode,
      reuse_scope: reuseScope,
      failure: failureMode,
    })}]->(${outerWhip})`
  );
  lines.push(`(${outerWhip})-[:PROCESS]->(${outerTrace})`);
  lines.push(
    `(${outerWhip})-[:VENT${renderEdgeProperties({
      condition: ventCondition,
      corridor: outerCorridor,
      repair_debt: '0',
      drift_coefficient: '0',
    })}]->(${outerVoid})`
  );
  lines.push(`(${outerTrace})-[:PROCESS]->(${primitive.id})`);
  return lines;
}

function expandWallingtonRotationPrimitive(
  primitive: StructuredPrimitive
): readonly string[] {
  let stages = parsePositiveInt(primitive.properties.stages, 4);
  let chunks = parsePositiveInt(primitive.properties.chunks, stages);
  
  if (primitive.properties.parameters) {
    const numParameters = parseNonnegativeInt(primitive.properties.parameters, 0);
    const dimension = resolveWallingtonDimension(numParameters);
    const meshConfig = GNOSTIC_MESHES[dimension];
    chunks = Math.max(1, meshConfig.peers);
    stages = Math.max(1, meshConfig.hops);
  }

  const chunkLabel = primitive.properties.chunkLabel ?? 'EncoderChunk';
  const stageLabel = primitive.properties.stageLabel ?? 'EncoderShard';
  const outputLabel = primitive.properties.outputLabel ?? 'FlowFrame';
  const alignmentLabel = primitive.properties.alignmentLabel ?? 'FlowFrame';
  const schedule = primitive.properties.schedule ?? 'wallington_rotation';
  const foldStrategy = primitive.properties.foldStrategy ?? 'chunk-order';
  const ingressRole =
    primitive.properties.ingressRole ?? 'wallington-rotation-ingress';
  const alignmentRole =
    primitive.properties.alignmentRole ?? 'stage-aligned-chunks';
  const outputRole = primitive.properties.outputRole ?? 'rotated-output';
  const rotationRole =
    primitive.properties.rotationRole ?? 'wallington-rotation';
  const stageParameters = String(
    parsePositiveInt(
      primitive.properties.stageParameters ?? primitive.properties.parameters,
      1
    )
  );
  const chunkNames = resolveNameList(
    primitive.properties.chunkNames,
    chunks,
    (index) => `${primitive.id}__chunk_${index}`
  );
  const stageNames = resolveNameList(
    primitive.properties.stageNames,
    stages,
    (index) => `${primitive.id}__stage_${index}`
  );

  const lines: string[] = [];
  lines.push(
    `// WallingtonRotation '${primitive.id}' expanded into chunked stage-pipeline topology.`
  );
  lines.push(
    renderNode(`${primitive.id}__ingress`, 'FlowFrame', {
      role: ingressRole,
      primitive: 'WallingtonRotation',
    })
  );
  lines.push(
    renderNode(`${primitive.id}__scheduler`, 'RotationScheduler', {
      schedule,
      stages: String(stages),
      chunks: String(chunks),
      role: rotationRole,
    })
  );
  for (const [chunkIndex, chunkName] of chunkNames.entries()) {
    lines.push(
      renderNode(chunkName, chunkLabel, {
        ordinal: String(chunkIndex),
      })
    );
  }
  lines.push(
    renderNode(`${primitive.id}__stage_aligned`, alignmentLabel, {
      role: alignmentRole,
      primitive: 'WallingtonRotation',
    })
  );
  for (const [stageIndex, stageName] of stageNames.entries()) {
    lines.push(
      renderNode(stageName, stageLabel, {
        stage: stageName,
        parameters: stageParameters,
        stage_index: String(stageIndex),
      })
    );
  }
  lines.push(
    renderNode(primitive.id, outputLabel, {
      primitive: 'WallingtonRotation',
      schedule,
      stages: String(stages),
      chunks: String(chunks),
      role: outputRole,
    })
  );
  lines.push(
    `(${primitive.id}__ingress)-[:PROCESS]->(${primitive.id}__scheduler)`
  );
  lines.push(
    `(${
      primitive.id
    }__scheduler)-[:FORK { schedule: '${schedule}' }]->(${chunkNames.join(
      ' | '
    )})`
  );
  lines.push(
    `(${chunkNames.join(' | ')})-[:FOLD { strategy: '${foldStrategy}' }]->(${
      primitive.id
    }__stage_aligned)`
  );
  lines.push(`(${primitive.id}__stage_aligned)-[:PROCESS]->(${stageNames[0]})`);
  for (let stageIndex = 0; stageIndex < stageNames.length - 1; stageIndex++) {
    lines.push(
      `(${stageNames[stageIndex]})-[:PROCESS]->(${stageNames[stageIndex + 1]})`
    );
  }
  lines.push(
    `(${stageNames[stageNames.length - 1]})-[:PROCESS]->(${primitive.id})`
  );

  return lines;
}

function expandWorthingtonWhipPrimitive(
  primitive: StructuredPrimitive
): readonly string[] {
  const shardCount = parsePositiveInt(
    primitive.properties.shardCount ?? primitive.properties.shards,
    2
  );
  const activeShards = Math.min(
    shardCount,
    parsePositiveInt(
      primitive.properties.activeShards ??
        primitive.properties.activePaths ??
        primitive.properties.active_paths,
      shardCount
    )
  );
  const stages = parsePositiveInt(primitive.properties.stages, 2);
  const chunks = parsePositiveInt(primitive.properties.chunks, stages);
  const shardNames = resolveNameList(
    primitive.properties.shardNames,
    shardCount,
    (index) => `${primitive.id}__shard_${index}`
  );
  const stageNameBases = parseNameList(primitive.properties.stageNames);
  const chunkLabel = primitive.properties.chunkLabel ?? 'EncoderChunk';
  const stageLabel = primitive.properties.stageLabel ?? 'EncoderShard';
  const schedule = primitive.properties.schedule ?? 'wallington_rotation';
  const outputLabel = primitive.properties.outputLabel ?? 'FlowFrame';
  const outputRole = primitive.properties.outputRole ?? 'collapsed-output';
  const ingressRole =
    primitive.properties.ingressRole ?? 'worthington-whip-ingress';
  const routerRole =
    primitive.properties.routerRole ?? 'worthington-whip-router';
  const collapseRole = primitive.properties.collapseRole ?? 'worthington-whip';
  const collapseBoundary = primitive.properties.collapseBoundary ?? 'shards';
  const collapseStrategy =
    primitive.properties.collapseStrategy ?? 'worthington_whip';
  const stageParameters = String(
    parsePositiveInt(
      primitive.properties.stageParameters ?? primitive.properties.parameters,
      1
    )
  );

  const lines: string[] = [];
  lines.push(
    `// WorthingtonWhip '${primitive.id}' expanded into routed shard-local Wallington rotations plus one collapse boundary.`
  );
  lines.push(
    renderNode(`${primitive.id}__ingress`, 'FlowFrame', {
      role: ingressRole,
      primitive: 'WorthingtonWhip',
    })
  );
  lines.push(
    renderNode(`${primitive.id}__router`, 'RoutingGate', {
      transformerlets: String(shardCount),
      active_paths: String(activeShards),
      role: routerRole,
    })
  );
  lines.push(
    renderNode(`${primitive.id}__collapse`, 'FoldOperator', {
      role: collapseRole,
      boundary: collapseBoundary,
      strategy: collapseStrategy,
    })
  );
  lines.push(
    renderNode(primitive.id, outputLabel, {
      primitive: 'WorthingtonWhip',
      shards: String(shardCount),
      active_shards: String(activeShards),
      role: outputRole,
    })
  );
  lines.push(
    `(${primitive.id}__ingress)-[:PROCESS]->(${primitive.id}__router)`
  );
  lines.push(`(${primitive.id}__router)-[:FORK]->(${shardNames.join(' | ')})`);

  const shardRotationNames: string[] = [];
  for (const [shardIndex, shardName] of shardNames.entries()) {
    const rotationName = `${shardName}__rotation`;
    const stageNames =
      stageNameBases.length === stages
        ? stageNameBases.map((stageName) => `${shardName}__${stageName}`)
        : Array.from(
            { length: stages },
            (_, stageOffset) => `${shardName}__stage_${stageOffset}`
          );
    shardRotationNames.push(rotationName);

    lines.push(
      renderNode(shardName, 'FlowFrame', {
        role: 'worthington-shard',
        shard: String(shardIndex),
      })
    );
    lines.push(
      `(${rotationName}: WallingtonRotation { stages: '${stages}', chunks: '${chunks}', schedule: '${schedule}', chunkLabel: '${chunkLabel}', stageLabel: '${stageLabel}', stageParameters: '${stageParameters}', stageNames: '[${stageNames.join(
        ','
      )}]', outputLabel: 'FlowFrame', outputRole: 'worthington-shard-output', ingressRole: 'worthington-shard-ingress', alignmentRole: 'worthington-shard-aligned', rotationRole: 'worthington-shard-rotation' })`
    );
    lines.push(`(${shardName})-[:PROCESS]->(${rotationName})`);
  }

  lines.push(
    `(${shardRotationNames.join(
      ' | '
    )})-[:FOLD { strategy: '${collapseStrategy}', boundary: '${collapseBoundary}' }]->(${
      primitive.id
    }__collapse)`
  );
  lines.push(`(${primitive.id}__collapse)-[:PROCESS]->(${primitive.id})`);

  return lines;
}

function expandHeteroMoAFabricPrimitive(
  primitive: StructuredPrimitive
): readonly string[] {
  let cpuLanes = parseNonnegativeInt(
    readAliasedProperty(primitive.properties, 'cpuLanes', 'cpu_lanes'),
    1
  );
  const gpuLanes = parseNonnegativeInt(
    readAliasedProperty(primitive.properties, 'gpuLanes', 'gpu_lanes'),
    0
  );
  const npuLanes = parseNonnegativeInt(
    readAliasedProperty(primitive.properties, 'npuLanes', 'npu_lanes'),
    0
  );
  const wasmLanes = parseNonnegativeInt(
    readAliasedProperty(
      primitive.properties,
      'wasmLanes',
      'wasm_lanes',
      'browserLanes',
      'browser_lanes'
    ),
    0
  );
  if (cpuLanes + gpuLanes + npuLanes + wasmLanes === 0) {
    cpuLanes = 1;
  }

  const blocks = parsePositiveInt(primitive.properties.blocks, 1);
  const activeBlocks = Math.min(
    blocks,
    parsePositiveInt(
      primitive.properties.activeBlocks ?? primitive.properties.active_blocks,
      blocks
    )
  );
  const heads = parsePositiveInt(primitive.properties.heads, 1);
  const activeHeads = Math.min(
    heads,
    parsePositiveInt(
      primitive.properties.activeHeads ?? primitive.properties.active_heads,
      heads
    )
  );
  const stages = parsePositiveInt(primitive.properties.stages, 1);
  const chunks = parsePositiveInt(primitive.properties.chunks, 1);
  const pairedKernel =
    readAliasedProperty(
      primitive.properties,
      'pairedKernel',
      'paired_kernel'
    ) ?? 'mirror';
  const scheduleStrategy =
    readAliasedProperty(
      primitive.properties,
      'scheduleStrategy',
      'schedule_strategy'
    ) ?? 'cannon';
  const launchGate =
    readAliasedProperty(primitive.properties, 'launchGate', 'launch_gate') ??
    'armed';
  const hedgeDelayTicks = String(
    parsePositiveInt(
      readAliasedProperty(
        primitive.properties,
        'hedgeDelayTicks',
        'hedge_delay_ticks'
      ),
      1
    )
  );
  const hedgePolicy =
    readAliasedProperty(primitive.properties, 'hedgePolicy', 'hedge_policy') ??
    'race';
  const frameProtocol =
    readAliasedProperty(
      primitive.properties,
      'frameProtocol',
      'frame_protocol'
    ) ?? 'aeon-10-byte-binary';
  const corridorBase =
    readAliasedProperty(primitive.properties, 'corridor', 'corridor_base') ??
    primitive.id;
  const corridorMode =
    readAliasedProperty(
      primitive.properties,
      'corridorMode',
      'corridor_mode'
    ) ?? 'readwrite';
  const reuseScope =
    readAliasedProperty(primitive.properties, 'reuseScope', 'reuse_scope') ??
    'corridor';
  const failureMode = primitive.properties.failure ?? 'vent';
  const traceLabel =
    readAliasedProperty(primitive.properties, 'traceLabel', 'trace_label') ??
    'CorridorTrace';
  const ventLabel =
    readAliasedProperty(primitive.properties, 'ventLabel', 'vent_label') ??
    'VoidBoundary';
  const ventCondition =
    readAliasedProperty(
      primitive.properties,
      'ventCondition',
      'vent_condition'
    ) ?? 'corridor-reject';
  const observableKind =
    readAliasedProperty(
      primitive.properties,
      'observableKind',
      'observable_kind'
    ) ?? 'queue-depth';
  const observable = primitive.properties.observable ?? 'beta1';
  const observableScale =
    readAliasedProperty(
      primitive.properties,
      'observableScale',
      'observable_scale'
    ) ?? '1';
  const observableOffset =
    readAliasedProperty(
      primitive.properties,
      'observableOffset',
      'observable_offset'
    ) ?? '0';
  const driftGap =
    readAliasedProperty(primitive.properties, 'driftGap', 'drift_gap') ?? '1';

  const backends = [
    { name: 'cpu', lanes: cpuLanes, role: 'cpu-control-helix' },
    { name: 'gpu', lanes: gpuLanes, role: 'gpu-wave-helix' },
    { name: 'npu', lanes: npuLanes, role: 'npu-route-helix' },
    { name: 'wasm', lanes: wasmLanes, role: 'wasm-browser-helix' },
  ].filter((backend) => backend.lanes > 0);

  const outputProperties = renderPrimitiveProperties({
    primitive: 'HeteroMoAFabric',
    cpu_lanes: String(cpuLanes),
    gpu_lanes: String(gpuLanes),
    npu_lanes: String(npuLanes),
    wasm_lanes: String(wasmLanes),
    paired_kernel: pairedKernel,
    schedule_strategy: scheduleStrategy,
    launch_gate: launchGate,
    hedge_delay_ticks: hedgeDelayTicks,
    hedge_policy: hedgePolicy,
    frame_protocol: frameProtocol,
    active_layers: String(backends.length),
  });

  const lines: string[] = [];
  lines.push(
    `// HeteroMoAFabric '${primitive.id}' expanded into backend-diverse mirrored MoA lanes with a cannon/helix meta race and laminar stream collapse.`
  );
  lines.push(
    `(${primitive.id}__ingress: FlowFrame { role: 'hetero-moa-ingress', primitive: 'HeteroMoAFabric' })`
  );
  lines.push(
    `(${primitive.id}__launch_gate: RoutingGate { role: 'paired-launch-gate', gate: '${launchGate}', schedule: '${scheduleStrategy}', frame_protocol: '${frameProtocol}' })`
  );
  lines.push(
    `(${
      primitive.id
    }__meta_scheduler: RotationScheduler { schedule: '${scheduleStrategy}', stages: '${Math.max(
      1,
      backends.length
    )}', chunks: '${Math.max(1, backends.length)}', role: 'meta-helix-race' })`
  );
  lines.push(
    `(${primitive.id}__global_collapse: FoldOperator { role: 'meta-layer-collapse', boundary: 'meta-race', strategy: '${scheduleStrategy}' })`
  );
  lines.push(
    renderNode(`${primitive.id}__global_trace`, traceLabel, {
      role: 'meta-layer-trace',
      corridor: `${corridorBase}/meta`,
      boundary: 'meta-race',
      primitive: 'HeteroMoAFabric',
    })
  );
  lines.push(
    renderNode(`${primitive.id}__global_void`, ventLabel, {
      role: 'meta-layer-void',
      corridor: `${corridorBase}/meta`,
      boundary: 'meta-race',
      primitive: 'HeteroMoAFabric',
    })
  );
  lines.push(`(${primitive.id}: FlowFrame${outputProperties})`);
  lines.push(
    `(${primitive.id}__ingress)-[:PROCESS]->(${primitive.id}__launch_gate)`
  );
  lines.push(
    `(${primitive.id}__launch_gate)-[:PROCESS]->(${primitive.id}__meta_scheduler)`
  );
  lines.push(
    `(${
      primitive.id
    }__meta_scheduler)-[:FORK { schedule: '${scheduleStrategy}', gate: '${launchGate}', helix: 'meta' }]->(${backends
      .map((backend) => `${primitive.id}__${backend.name}__scheduler`)
      .join(' | ')})`
  );

  const layerOutputs: string[] = [];
  for (const backend of backends) {
    const schedulerId = `${primitive.id}__${backend.name}__scheduler`;
    const foldId = `${primitive.id}__${backend.name}__fold`;
    const traceId = `${primitive.id}__${backend.name}__trace`;
    const voidId = `${primitive.id}__${backend.name}__void`;
    const layerCorridor = `${corridorBase}/${backend.name}/layer`;

    lines.push(
      `(${schedulerId}: RotationScheduler { schedule: '${scheduleStrategy}', stages: '${backend.lanes}', chunks: '${backend.lanes}', role: '${backend.role}', device_affinity: '${backend.name}', cursor_mode: 'helix' })`
    );
    lines.push(
      `(${foldId}: FoldOperator { role: '${backend.name}-layer-collapse', boundary: '${backend.name}-layer', strategy: '${scheduleStrategy}' })`
    );
    lines.push(
      renderNode(traceId, traceLabel, {
        role: `${backend.name}-layer-trace`,
        corridor: layerCorridor,
        boundary: `${backend.name}-layer`,
        primitive: 'HeteroMoAFabric',
      })
    );
    lines.push(
      renderNode(voidId, ventLabel, {
        role: `${backend.name}-layer-void`,
        corridor: layerCorridor,
        boundary: `${backend.name}-layer`,
        primitive: 'HeteroMoAFabric',
      })
    );

    const pairOutputs: string[] = [];
    for (let laneIndex = 0; laneIndex < backend.lanes; laneIndex++) {
      const laneBase = `${primitive.id}__${backend.name}_${laneIndex}`;
      const primaryId = `${laneBase}__primary`;
      const shadowId = `${laneBase}__shadow`;
      const pairId = `${laneBase}__pair`;
      const pairTraceId = `${laneBase}__trace`;
      const pairVoidId = `${laneBase}__void`;
      const pairCorridor = `${corridorBase}/${backend.name}/lane-${laneIndex}/pair`;
      pairOutputs.push(pairTraceId);

      lines.push(
        `(${pairId}: RoutingGate { role: 'paired-kernel-adjudicator', backend: '${backend.name}', lane: '${laneIndex}', pair_kernel: '${pairedKernel}', hedge_delay_ticks: '${hedgeDelayTicks}', hedge_policy: '${hedgePolicy}', frame_protocol: '${frameProtocol}', device_affinity: '${backend.name}' })`
      );
      lines.push(
        renderNode(pairTraceId, traceLabel, {
          role: 'paired-kernel-trace',
          corridor: pairCorridor,
          boundary: 'paired-kernel',
          backend: backend.name,
          lane: String(laneIndex),
          primitive: 'HeteroMoAFabric',
        })
      );
      lines.push(
        renderNode(pairVoidId, ventLabel, {
          role: 'paired-kernel-void',
          corridor: pairCorridor,
          boundary: 'paired-kernel',
          backend: backend.name,
          lane: String(laneIndex),
          primitive: 'HeteroMoAFabric',
        })
      );
      lines.push(
        `(${primaryId}: StructuredMoA { blocks: '${blocks}', activeBlocks: '${activeBlocks}', heads: '${heads}', activeHeads: '${activeHeads}', stages: '${stages}', chunks: '${chunks}', corridor: '${corridorBase}/${backend.name}/lane-${laneIndex}/primary', failure: '${failureMode}', corridor_mode: '${corridorMode}', reuse_scope: '${reuseScope}', observable_kind: '${observableKind}', observable: '${observable}', observable_scale: '${observableScale}', observable_offset: '${observableOffset}', drift_gap: '${driftGap}', pair_role: 'primary', device_affinity: '${backend.name}', lane: '${laneIndex}' })`
      );
      lines.push(
        `(${shadowId}: StructuredMoA { blocks: '${blocks}', activeBlocks: '${activeBlocks}', heads: '${heads}', activeHeads: '${activeHeads}', stages: '${stages}', chunks: '${chunks}', corridor: '${corridorBase}/${backend.name}/lane-${laneIndex}/shadow', failure: '${failureMode}', corridor_mode: '${corridorMode}', reuse_scope: '${reuseScope}', observable_kind: '${observableKind}', observable: '${observable}', observable_scale: '${observableScale}', observable_offset: '${observableOffset}', drift_gap: '${driftGap}', pair_role: 'shadow', device_affinity: '${backend.name}', lane: '${laneIndex}' })`
      );
      lines.push(
        `(${schedulerId})-[:FORK { schedule: '${scheduleStrategy}', gate: '${launchGate}', backend: '${backend.name}', lane: '${laneIndex}' }]->(${primaryId} | ${shadowId})`
      );
      lines.push(
        `(${primaryId} | ${shadowId})-[:RACE${renderEdgeProperties({
          strategy: scheduleStrategy,
          boundary: 'paired-kernel',
          corridor: pairCorridor,
          corridor_mode: corridorMode,
          reuse_scope: reuseScope,
          failure: failureMode,
          hedge_delay_ticks: hedgeDelayTicks,
          hedge_policy: hedgePolicy,
          pair_kernel: pairedKernel,
          frame_protocol: frameProtocol,
          backend: backend.name,
          lane: String(laneIndex),
        })}]->(${pairId})`
      );
      lines.push(`(${pairId})-[:PROCESS]->(${pairTraceId})`);
      lines.push(
        `(${pairId})-[:VENT${renderEdgeProperties({
          condition: ventCondition,
          corridor: pairCorridor,
          repair_debt: '0',
          drift_coefficient: '0',
        })}]->(${pairVoidId})`
      );
    }

    lines.push(
      `(${pairOutputs.join(' | ')})-[:FOLD${renderEdgeProperties({
        strategy: scheduleStrategy,
        boundary: `${backend.name}-layer`,
        corridor: layerCorridor,
        corridor_mode: corridorMode,
        reuse_scope: reuseScope,
        failure: failureMode,
        device_affinity: backend.name,
      })}]->(${foldId})`
    );
    lines.push(`(${foldId})-[:PROCESS]->(${traceId})`);
    lines.push(
      `(${foldId})-[:VENT${renderEdgeProperties({
        condition: ventCondition,
        corridor: layerCorridor,
        repair_debt: '0',
        drift_coefficient: '0',
      })}]->(${voidId})`
    );
    layerOutputs.push(traceId);
  }

  lines.push(
    `(${layerOutputs.join(' | ')})-[:FOLD${renderEdgeProperties({
      strategy: scheduleStrategy,
      boundary: 'meta-race',
      corridor: `${corridorBase}/meta`,
      corridor_mode: corridorMode,
      reuse_scope: reuseScope,
      failure: failureMode,
      frame_protocol: frameProtocol,
    })}]->(${primitive.id}__global_collapse)`
  );
  lines.push(
    `(${primitive.id}__global_collapse)-[:PROCESS]->(${primitive.id}__global_trace)`
  );
  lines.push(
    `(${primitive.id}__global_collapse)-[:VENT${renderEdgeProperties({
      condition: ventCondition,
      corridor: `${corridorBase}/meta`,
      repair_debt: '0',
      drift_coefficient: '0',
    })}]->(${primitive.id}__global_void)`
  );
  lines.push(`(${primitive.id}__global_trace)-[:PROCESS]->(${primitive.id})`);

  return lines;
}

function expandStructuredPrimitive(
  primitive: StructuredPrimitive
): readonly string[] {
  switch (primitive.kind) {
    case 'StructuredMoA':
      return expandStructuredMoaPrimitive(primitive);
    case 'WallingtonRotation':
      return expandWallingtonRotationPrimitive(primitive);
    case 'WorthingtonWhip':
      return expandWorthingtonWhipPrimitive(primitive);
    case 'HeteroMoAFabric':
      return expandHeteroMoAFabricPrimitive(primitive);
  }
}

/**
 * Handles the gnosis expand Structured Primitives Source workflow.
 */
export function expandStructuredPrimitivesSource(source: string): string {
  let expandedSource = source;

  for (
    let passIndex = 0;
    passIndex < MAX_STRUCTURED_PRIMITIVE_PASSES;
    passIndex++
  ) {
    const units = collectStructuredPrimitiveUnits(expandedSource);
    const primitives = units
      .map((unit) => parseStructuredPrimitive(unit))
      .filter(
        (primitive): primitive is StructuredPrimitive => primitive !== null
      );

    if (primitives.length === 0) {
      return expandedSource;
    }

    const primitiveIds = new Set(primitives.map((primitive) => primitive.id));
    expandedSource = units
      .flatMap((unit) => {
        const primitive = parseStructuredPrimitive(unit);
        if (primitive) {
          return expandStructuredPrimitive(primitive);
        }

        return [rewriteIncomingPrimitiveTargets(unit, primitiveIds)];
      })
      .join('\n');
  }

  throw new Error(
    `Structured primitive expansion exceeded ${MAX_STRUCTURED_PRIMITIVE_PASSES} passes.`
  );
}
