// @a0n/.ignored_gnosis/src/structured-data.ts
export type StructuredDataFormat = 'jsonld' | 'microdata' | 'rdfa';

export type StructuredDataProvenance = 'declared' | 'derived' | 'speculative';

export interface StructuredDataSourceSpan {
  file: string;
  startLine: number;
  startColumn: number;
  endLine: number;
  endColumn: number;
}

export interface StructuredDataContextMap {
  [prefix: string]: string;
}

export interface StructuredDataDeclaration {
  sdType?: string;
  sdId?: string;
  sdProp?: string;
  sdRef?: string;
  sdValue?: string;
  sdLang?: string;
  sdContext?: StructuredDataContextMap;
  sdContextError?: string;
  textContent?: string;
  sourceSpan?: StructuredDataSourceSpan;
}

export type StructuredDataNodeRef =
  | { kind: 'iri'; iri: string }
  | { kind: 'blank'; id: string };

export interface StructuredDataLiteral {
  kind: 'literal';
  value: string;
  language?: string;
}

export interface StructuredDataTriple {
  subject: StructuredDataNodeRef;
  predicate: string;
  object: StructuredDataNodeRef | StructuredDataLiteral;
  provenance: StructuredDataProvenance;
  sourceSpan?: StructuredDataSourceSpan;
}

export interface StructuredDataGraph {
  routeId: string;
  context: StructuredDataContextMap;
  triples: StructuredDataTriple[];
  schemaTypes: string[];
  vocabularies: string[];
  tripleCount: number;
}

export interface StructuredDataGraphSummary {
  schemaTypes: string[];
  tripleCount: number;
  vocabularies: string[];
}

export type StructuredDataValidationCode =
  | 'SD_UNRESOLVED_REF'
  | 'SD_INCOMPATIBLE_DUPLICATE_ID'
  | 'SD_INVALID_CONTEXT_MAP'
  | 'SD_ILLEGAL_LITERAL_SHAPE'
  | 'SD_CONFLICTING_ROUTE_DECLARATIONS';

export interface StructuredDataValidationIssue {
  code: StructuredDataValidationCode;
  message: string;
  severity: 'warning' | 'error';
  sourceSpan?: StructuredDataSourceSpan;
}

export interface StructuredDataCompilationResult {
  graph: StructuredDataGraph;
  issues: StructuredDataValidationIssue[];
}

export interface StructuredDataRouteContext {
  routeId?: string;
  baseIri?: string;
  context?: StructuredDataContextMap;
  speculative?: boolean;
}

export interface StructuredDataEmitOptions {
  formats?: StructuredDataFormat[];
  compactContext?: boolean;
}

export interface StructuredDataEmitResult {
  jsonld?: string;
  rdfa?: string;
  microdata?: string;
  skippedMicrodataTriples: number;
}

export type StructuredProofSystem =
  | 'lean4'
  | 'aeon-logic'
  | 'tlaplus'
  | 'gg'
  | 'coq'
  | 'agda'
  | string;

export interface StructuredProofDeclaration {
  system: StructuredProofSystem;
  theorem: string;
  status?: 'verified' | 'sorry' | 'conjecture';
  source?: string;
  expression?: string;
  sourceSpan?: StructuredDataSourceSpan;
}

export interface StructuredProofAttrInput {
  theorem: string;
  system: StructuredProofSystem;
  status?: 'verified' | 'sorry' | 'conjecture';
  source?: string;
  expression?: string;
}

export interface StructuredDataCodemodPlan {
  pattern:
    | 'Organization'
    | 'PersonContactPoint'
    | 'FAQPage'
    | 'Recipe'
    | 'Product'
    | 'Event'
    | 'Article';
  declarations: StructuredDataDeclaration[];
}

const DEFAULT_CONTEXT: StructuredDataContextMap = {
  schema: 'https://schema.org/',
  rdf: 'http://www.w3.org/1999/02/22-rdf-syntax-ns#',
  rdfs: 'http://www.w3.org/2000/01/rdf-schema#',
  xsd: 'http://www.w3.org/2001/XMLSchema#',
  foaf: 'http://xmlns.com/foaf/0.1/',
};

const RDF_TYPE = 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type';
const SCHEMA_BASE = 'https://schema.org/';

/**
 * Handles the gnosis extract Structured Declarations workflow.
 */
export function extractStructuredDeclarations(
  source: string,
  filePath: string
): StructuredDataDeclaration[] {
  const declarations: StructuredDataDeclaration[] = [];
  const tokenRegex = /<\/?[A-Za-z][^>]*>/g;
  const stack: StructuredDataContextMap[] = [Object.freeze({ ...DEFAULT_CONTEXT })];

  for (const match of source.matchAll(tokenRegex)) {
    const token = match[0];
    const index = match.index ?? 0;
    const isCloseTag = token.startsWith('</');
    const isSelfClosing = /\/>\s*$/.test(token);

    if (isCloseTag) {
      if (stack.length > 1) {
        stack.pop();
      }
      continue;
    }

    const attrs = parseAttributes(token);
    const topContext = stack[stack.length - 1] ?? DEFAULT_CONTEXT;

    const localContextRaw = attrs['data-sd-context'];
    let localContext: StructuredDataContextMap | undefined;
    let contextError: string | undefined;
    if (localContextRaw !== undefined) {
      const parsed = parseContextMap(localContextRaw);
      if ('error' in parsed) {
        contextError = parsed.error;
      } else {
        localContext = parsed.context;
      }
    }

    const effectiveContext: StructuredDataContextMap = {
      ...topContext,
      ...(localContext ?? {}),
    };

    const hasStructuredEntityAttr = Object.keys(attrs).some(
      (key) => key.startsWith('data-sd-') && key !== 'data-sd-context'
    );
    if (hasStructuredEntityAttr || contextError !== undefined) {
      declarations.push({
        ...(attrs['data-sd-type'] !== undefined && { sdType: attrs['data-sd-type'] }),
        ...(attrs['data-sd-id'] !== undefined && { sdId: attrs['data-sd-id'] }),
        ...(attrs['data-sd-prop'] !== undefined && { sdProp: attrs['data-sd-prop'] }),
        ...(attrs['data-sd-ref'] !== undefined && { sdRef: attrs['data-sd-ref'] }),
        ...(attrs['data-sd-value'] !== undefined && { sdValue: attrs['data-sd-value'] }),
        ...(attrs['data-sd-lang'] !== undefined && { sdLang: attrs['data-sd-lang'] }),
        sdContext: effectiveContext,
        ...(contextError !== undefined && { sdContextError: contextError }),
        sourceSpan: sourceIndexToSpan(source, index, filePath),
      });
    }

    if (!isSelfClosing) {
      stack.push(effectiveContext);
    }
  }

  return declarations.sort(compareDeclarationOrder);
}

/**
 * Parses the Proof Declaration From Attributes.
 */
export function parseProofDeclarationFromAttributes(
  attributes: Record<string, string>,
  sourceSpan?: StructuredDataSourceSpan
): StructuredProofDeclaration | null {
  const raw = attributes['data-proof'];
  if (!raw) {
    return null;
  }

  const parts = raw.split(':');
  if (parts.length < 2) {
    return null;
  }

  const system = parts.shift()?.trim();
  const theorem = parts.join(':').trim();
  if (!system || !theorem) {
    return null;
  }

  const status = attributes['data-proof-status'];
  const typedStatus =
    status === 'verified' || status === 'sorry' || status === 'conjecture'
      ? status
      : undefined;

  return {
    system,
    theorem,
    ...(typedStatus !== undefined && { status: typedStatus }),
    ...(attributes['data-proof-source'] !== undefined && { source: attributes['data-proof-source'] }),
    ...(attributes['data-proof-expr'] !== undefined && { expression: attributes['data-proof-expr'] }),
    ...(sourceSpan !== undefined && { sourceSpan }),
  };
}

/**
 * Parses the Proof Declaration From Value.
 */
export function parseProofDeclarationFromValue(
  rawProof: string,
  options?: {
    status?: 'verified' | 'sorry' | 'conjecture';
    source?: string;
    expression?: string;
    sourceSpan?: StructuredDataSourceSpan;
  }
): StructuredProofDeclaration | null {
  return parseProofDeclarationFromAttributes(
    {
      'data-proof': rawProof,
      ...(options?.status ? { 'data-proof-status': options.status } : {}),
      ...(options?.source ? { 'data-proof-source': options.source } : {}),
      ...(options?.expression ? { 'data-proof-expr': options.expression } : {}),
    },
    options?.sourceSpan
  );
}

/**
 * Handles the gnosis proof Attr workflow.
 */
export function proofAttr(input: StructuredProofAttrInput): string {
  const attrs: string[] = [
    `data-proof="${escapeAttr(`${input.system}:${input.theorem}`)}"`,
  ];
  if (input.status) {
    attrs.push(`data-proof-status="${escapeAttr(input.status)}"`);
  }
  if (input.source) {
    attrs.push(`data-proof-source="${escapeAttr(input.source)}"`);
  }
  if (input.expression) {
    attrs.push(`data-proof-expr="${escapeAttr(input.expression)}"`);
  }
  return attrs.join(' ');
}

/**
 * Handles the gnosis compile Structured Graph workflow.
 */
export function compileStructuredGraph(
  declarations: StructuredDataDeclaration[],
  routeContext: StructuredDataRouteContext = {}
): StructuredDataCompilationResult {
  const sortedDeclarations = [...declarations].sort(compareDeclarationOrder);
  const context: StructuredDataContextMap = {
    ...DEFAULT_CONTEXT,
    ...(routeContext.context ?? {}),
  };
  const issues: StructuredDataValidationIssue[] = [];
  const routeId = routeContext.routeId ?? '/';
  const triples: StructuredDataTriple[] = [];
  const knownNodeIds = new Set<string>();
  const pendingRefs: Array<{
    ref: string;
    tripleIndex: number;
    declaration: StructuredDataDeclaration;
  }> = [];
  const idLiteralFingerprint = new Map<string, string>();

  for (const declaration of sortedDeclarations) {
    if (declaration.sdContextError) {
      issues.push({
        code: 'SD_INVALID_CONTEXT_MAP',
        message: `Invalid data-sd-context JSON: ${declaration.sdContextError}`,
        severity: 'error',
        ...(declaration.sourceSpan !== undefined && { sourceSpan: declaration.sourceSpan }),
      });
    }

    const effectiveContext = {
      ...context,
      ...(declaration.sdContext ?? {}),
    };

    const subject = resolveSubject(declaration, routeContext, effectiveContext);
    if (subject.kind === 'iri') {
      knownNodeIds.add(subject.iri);
    } else {
      knownNodeIds.add(`_:${subject.id}`);
    }

    if (declaration.sdType) {
      const expandedType = expandIri(
        declaration.sdType,
        effectiveContext,
        routeContext.baseIri
      );
      triples.push({
        subject,
        predicate: RDF_TYPE,
        object: { kind: 'iri', iri: expandedType },
        provenance: 'declared',
        ...(declaration.sourceSpan !== undefined && { sourceSpan: declaration.sourceSpan }),
      });
    }

    if (declaration.sdProp) {
      const predicate = expandIri(
        declaration.sdProp,
        effectiveContext,
        routeContext.baseIri
      );

      if (declaration.sdRef) {
        const refIri = expandIri(
          declaration.sdRef,
          effectiveContext,
          routeContext.baseIri
        );
        const triple: StructuredDataTriple = {
          subject,
          predicate,
          object: { kind: 'iri', iri: refIri },
          provenance: 'declared',
          ...(declaration.sourceSpan !== undefined && { sourceSpan: declaration.sourceSpan }),
        };
        triples.push(triple);
        pendingRefs.push({
          ref: refIri,
          tripleIndex: triples.length - 1,
          declaration,
        });
      } else {
        const rawLiteral =
          declaration.sdValue ?? declaration.textContent ?? declaration.sdId;
        if (!rawLiteral || rawLiteral.trim().length === 0) {
          issues.push({
            code: 'SD_ILLEGAL_LITERAL_SHAPE',
            message: `Literal value is required for ${declaration.sdProp} when data-sd-ref is not provided.`,
            severity: 'error',
            ...(declaration.sourceSpan !== undefined && { sourceSpan: declaration.sourceSpan }),
          });
        } else {
          if (/^\s*[[{]/.test(rawLiteral.trim())) {
            issues.push({
              code: 'SD_ILLEGAL_LITERAL_SHAPE',
              message:
                'Literal appears to contain JSON object/array text. Use data-sd-ref or scalar literal values.',
              severity: 'warning',
              ...(declaration.sourceSpan !== undefined && { sourceSpan: declaration.sourceSpan }),
            });
          }
          const literalFingerprintKey = `${nodeKey(subject)}|${predicate}`;
          const prev = idLiteralFingerprint.get(literalFingerprintKey);
          if (prev && prev !== rawLiteral) {
            issues.push({
              code: 'SD_INCOMPATIBLE_DUPLICATE_ID',
              message:
                'Duplicate declaration assigns conflicting literal values to the same subject/predicate pair.',
              severity: 'error',
              ...(declaration.sourceSpan !== undefined && { sourceSpan: declaration.sourceSpan }),
            });
          }
          idLiteralFingerprint.set(literalFingerprintKey, rawLiteral);
          triples.push({
            subject,
            predicate,
            object: {
              kind: 'literal',
              value: rawLiteral,
              ...(declaration.sdLang !== undefined && { language: declaration.sdLang }),
            },
            provenance: 'declared',
            ...(declaration.sourceSpan !== undefined && { sourceSpan: declaration.sourceSpan }),
          });
        }
      }
    }
  }

  for (const pendingRef of pendingRefs) {
    if (!knownNodeIds.has(pendingRef.ref)) {
      issues.push({
        code: 'SD_UNRESOLVED_REF',
        message: `Reference '${pendingRef.ref}' does not resolve to a declared node ID in this route scope.`,
        severity: 'error',
        ...(pendingRef.declaration.sourceSpan !== undefined && { sourceSpan: pendingRef.declaration.sourceSpan }),
      });
      const triple = triples[pendingRef.tripleIndex];
      if (triple) {
        triple.provenance = 'derived';
      }
    }
  }

  if (routeContext.speculative) {
    const subjectsWithName = new Set<string>();
    for (const triple of triples) {
      if (
        triple.predicate === `${SCHEMA_BASE}name` &&
        triple.object.kind === 'literal'
      ) {
        subjectsWithName.add(nodeKey(triple.subject));
      }
    }

    for (const triple of triples) {
      if (triple.predicate !== RDF_TYPE || triple.object.kind !== 'iri') {
        continue;
      }
      const key = nodeKey(triple.subject);
      if (subjectsWithName.has(key)) {
        continue;
      }
      const speculativeName = deriveSpeculativeName(triple.object.iri);
      if (!speculativeName) {
        continue;
      }
      triples.push({
        subject: triple.subject,
        predicate: `${SCHEMA_BASE}name`,
        object: { kind: 'literal', value: speculativeName },
        provenance: 'speculative',
        ...(triple.sourceSpan !== undefined && { sourceSpan: triple.sourceSpan }),
      });
      subjectsWithName.add(key);
    }
  }

  const routeRootTypeSet = new Set<string>();
  for (const triple of triples) {
    if (triple.predicate === RDF_TYPE && triple.object.kind === 'iri') {
      routeRootTypeSet.add(triple.object.iri);
    }
  }
  if (routeRootTypeSet.size > 24) {
    issues.push({
      code: 'SD_CONFLICTING_ROUTE_DECLARATIONS',
      message:
        'Route declares an unusually large number of root types. Verify route-level declarations are not conflicting.',
      severity: 'warning',
    });
  }

  const schemaTypes = Array.from(
    new Set(
      triples
        .filter((triple) => triple.predicate === RDF_TYPE)
        .map((triple) =>
          triple.object.kind === 'iri' ? triple.object.iri : undefined
        )
        .filter((value): value is string => typeof value === 'string')
    )
  ).sort();

  const vocabularies = Array.from(
    new Set(
      triples
        .map((triple) => vocabularyOfIri(triple.predicate))
        .filter((value): value is string => value !== null)
    )
  ).sort();

  const graph: StructuredDataGraph = {
    routeId,
    context,
    triples,
    schemaTypes,
    vocabularies,
    tripleCount: triples.length,
  };

  return { graph, issues };
}

/**
 * Handles the gnosis emit Structured Data workflow.
 */
export function emitStructuredData(
  graph: StructuredDataGraph,
  options: StructuredDataEmitOptions = {}
): StructuredDataEmitResult {
  const formats = options.formats ?? ['jsonld'];
  let jsonld: string | undefined;
  let rdfa: string | undefined;
  let microdata: string | undefined;
  let skippedMicrodataTriples = 0;

  if (formats.includes('jsonld')) {
    const compact = options.compactContext !== false;
    const nodeMap = new Map<string, Record<string, unknown>>();
    for (const triple of graph.triples) {
      const subjectKey = nodeKey(triple.subject);
      const subjectNode = nodeMap.get(subjectKey) ?? { '@id': serializeNodeRef(triple.subject) };
      const predicate = compact
        ? compactIri(triple.predicate, graph.context)
        : triple.predicate;

      if (triple.predicate === RDF_TYPE && triple.object.kind === 'iri') {
        const typeValue = compact
          ? compactIri(triple.object.iri, graph.context)
          : triple.object.iri;
        pushJsonLdValue(subjectNode, '@type', typeValue);
      } else {
        const objectValue =
          triple.object.kind === 'literal'
            ? literalToJsonLd(triple.object)
            : { '@id': serializeNodeRef(triple.object) };
        pushJsonLdValue(subjectNode, predicate, objectValue);
      }

      nodeMap.set(subjectKey, subjectNode);
    }

    const graphObject: Record<string, unknown> = {
      ...(compact ? { '@context': graph.context } : {}),
      '@graph': Array.from(nodeMap.values()),
    };
    jsonld = JSON.stringify(graphObject, null, 2);
  }

  if (formats.includes('rdfa')) {
    const lines: string[] = ['<div vocab="https://schema.org/">'];
    const grouped = groupTriplesBySubject(graph.triples);
    for (const [subject, triples] of grouped.entries()) {
      lines.push(`  <div about="${escapeAttr(subject)}">`);
      for (const triple of triples) {
        if (triple.predicate === RDF_TYPE && triple.object.kind === 'iri') {
          lines.push(
            `    <span typeof="${escapeAttr(triple.object.iri)}"></span>`
          );
          continue;
        }

        if (triple.object.kind === 'literal') {
          lines.push(
            `    <meta property="${escapeAttr(triple.predicate)}" content="${escapeAttr(triple.object.value)}"${
              triple.object.language
                ? ` lang="${escapeAttr(triple.object.language)}"`
                : ''
            } />`
          );
        } else {
          lines.push(
            `    <link property="${escapeAttr(triple.predicate)}" href="${escapeAttr(serializeNodeRef(triple.object))}" />`
          );
        }
      }
      lines.push('  </div>');
    }
    lines.push('</div>');
    rdfa = lines.join('\n');
  }

  if (formats.includes('microdata')) {
    const lines: string[] = [];
    const grouped = groupTriplesBySubject(graph.triples);

    for (const [subject, triples] of grouped.entries()) {
      const typeTriple = triples.find(
        (triple) => triple.predicate === RDF_TYPE && triple.object.kind === 'iri'
      );
      if (!typeTriple || typeTriple.object.kind !== 'iri') {
        skippedMicrodataTriples += triples.length;
        continue;
      }
      const microType = toMicrodataType(typeTriple.object.iri);
      if (!microType) {
        skippedMicrodataTriples += triples.length;
        continue;
      }

      lines.push(
        `<div itemscope itemtype="${escapeAttr(microType)}" itemid="${escapeAttr(subject)}">`
      );
      for (const triple of triples) {
        if (triple.predicate === RDF_TYPE) {
          continue;
        }
        const itemprop = toMicrodataProp(triple.predicate);
        if (!itemprop) {
          skippedMicrodataTriples += 1;
          continue;
        }
        if (triple.object.kind === 'literal') {
          lines.push(
            `  <meta itemprop="${escapeAttr(itemprop)}" content="${escapeAttr(triple.object.value)}" />`
          );
        } else {
          lines.push(
            `  <link itemprop="${escapeAttr(itemprop)}" href="${escapeAttr(serializeNodeRef(triple.object))}" />`
          );
        }
      }
      lines.push('</div>');
    }

    microdata = lines.join('\n');
  }

  return {
    ...(jsonld !== undefined && { jsonld }),
    ...(rdfa !== undefined && { rdfa }),
    ...(microdata !== undefined && { microdata }),
    skippedMicrodataTriples,
  };
}

/**
 * Handles the gnosis summarize Structured Data Graph workflow.
 */
export function summarizeStructuredDataGraph(
  graph: StructuredDataGraph
): StructuredDataGraphSummary {
  return {
    schemaTypes: [...graph.schemaTypes],
    tripleCount: graph.tripleCount,
    vocabularies: [...graph.vocabularies],
  };
}

/**
 * Builds the Schema Codemod Plans.
 */
export function buildSchemaCodemodPlans(): StructuredDataCodemodPlan[] {
  return [
    {
      pattern: 'Organization',
      declarations: [
        { sdType: 'schema:Organization', sdId: '#organization' },
        {
          sdId: '#organization',
          sdProp: 'schema:name',
          sdValue: 'Your Organization',
        },
        {
          sdId: '#organization',
          sdProp: 'schema:url',
          sdValue: 'https://example.com',
        },
      ],
    },
    {
      pattern: 'PersonContactPoint',
      declarations: [
        { sdType: 'schema:Person', sdId: '#person' },
        { sdId: '#person', sdProp: 'schema:name', sdValue: 'Jane Doe' },
        {
          sdType: 'schema:ContactPoint',
          sdId: '#contact',
          sdProp: 'schema:email',
          sdValue: 'contact@example.com',
        },
        {
          sdId: '#person',
          sdProp: 'schema:contactPoint',
          sdRef: '#contact',
        },
      ],
    },
    {
      pattern: 'FAQPage',
      declarations: [
        { sdType: 'schema:FAQPage', sdId: '#faq' },
        { sdType: 'schema:Question', sdId: '#faq-q1' },
        { sdId: '#faq-q1', sdProp: 'schema:name', sdValue: 'Question?' },
        {
          sdType: 'schema:Answer',
          sdId: '#faq-a1',
          sdProp: 'schema:text',
          sdValue: 'Answer.',
        },
        {
          sdId: '#faq-q1',
          sdProp: 'schema:acceptedAnswer',
          sdRef: '#faq-a1',
        },
        { sdId: '#faq', sdProp: 'schema:mainEntity', sdRef: '#faq-q1' },
      ],
    },
    {
      pattern: 'Recipe',
      declarations: [
        { sdType: 'schema:Recipe', sdId: '#recipe' },
        { sdId: '#recipe', sdProp: 'schema:name', sdValue: 'Recipe Name' },
        {
          sdId: '#recipe',
          sdProp: 'schema:recipeIngredient',
          sdValue: '1 cup ingredient',
        },
      ],
    },
    {
      pattern: 'Product',
      declarations: [
        { sdType: 'schema:Product', sdId: '#product' },
        { sdId: '#product', sdProp: 'schema:name', sdValue: 'Product Name' },
        { sdType: 'schema:Offer', sdId: '#offer' },
        { sdId: '#offer', sdProp: 'schema:price', sdValue: '29.99' },
        {
          sdId: '#offer',
          sdProp: 'schema:priceCurrency',
          sdValue: 'USD',
        },
        { sdId: '#product', sdProp: 'schema:offers', sdRef: '#offer' },
      ],
    },
    {
      pattern: 'Event',
      declarations: [
        { sdType: 'schema:Event', sdId: '#event' },
        { sdId: '#event', sdProp: 'schema:name', sdValue: 'Event Name' },
        {
          sdId: '#event',
          sdProp: 'schema:startDate',
          sdValue: '2026-01-01T00:00:00Z',
        },
      ],
    },
    {
      pattern: 'Article',
      declarations: [
        { sdType: 'schema:Article', sdId: '#article' },
        { sdId: '#article', sdProp: 'schema:headline', sdValue: 'Headline' },
        {
          sdId: '#article',
          sdProp: 'schema:datePublished',
          sdValue: '2026-01-01',
        },
      ],
    },
  ];
}

/**
 * Handles the gnosis json Ld To Declarations workflow.
 */
export function jsonLdToDeclarations(
  input: Record<string, unknown> | Record<string, unknown>[]
): StructuredDataDeclaration[] {
  const declarations: StructuredDataDeclaration[] = [];
  const nodes = Array.isArray(input) ? input : [input];
  let autoNode = 0;

  const visitNode = (
    node: Record<string, unknown>,
    parentId?: string,
    parentPredicate?: string
  ): string => {
    const nodeIdRaw = valueToString(node['@id']) ?? `#jsonld-node-${autoNode++}`;
    const nodeId = nodeIdRaw.startsWith('#') ? nodeIdRaw : `#${nodeIdRaw}`;
    const contextValue = node['@context'];
    const context =
      contextValue && typeof contextValue === 'object' && !Array.isArray(contextValue)
        ? (contextValue as Record<string, string>)
        : undefined;
    const typeValue = node['@type'];
    const normalizedTypes = Array.isArray(typeValue)
      ? typeValue.map((value) => normalizeSemanticType(String(value)))
      : typeValue
        ? [normalizeSemanticType(String(typeValue))]
        : [];
    if (normalizedTypes.length > 0) {
      for (const sdType of normalizedTypes) {
        declarations.push({
          sdId: nodeId,
          sdType,
          ...(context ? { sdContext: context } : {}),
        });
      }
    } else {
      declarations.push({
        sdId: nodeId,
        sdType: 'schema:Thing',
        ...(context ? { sdContext: context } : {}),
      });
    }

    if (parentId && parentPredicate) {
      declarations.push({
        sdId: parentId,
        sdProp: normalizeSemanticType(parentPredicate),
        sdRef: nodeId,
      });
    }

    for (const [key, value] of Object.entries(node)) {
      if (key === '@context' || key === '@id' || key === '@type') {
        continue;
      }
      const predicate = normalizeSemanticType(key);
      const values = Array.isArray(value) ? value : [value];
      for (const item of values) {
        if (item === null || item === undefined) {
          continue;
        }
        if (typeof item === 'object') {
          const typedObject = item as Record<string, unknown>;
          if (typeof typedObject['@value'] === 'string') {
            declarations.push({
              sdId: nodeId,
              sdProp: predicate,
              sdValue: typedObject['@value'],
              ...(typeof typedObject['@language'] === 'string' && { sdLang: typedObject['@language'] }),
            });
          } else {
            visitNode(typedObject, nodeId, predicate);
          }
          continue;
        }
        declarations.push({
          sdId: nodeId,
          sdProp: predicate,
          sdValue: String(item),
        });
      }
    }

    return nodeId;
  };

  for (const node of nodes) {
    visitNode(node);
  }

  return declarations;
}

function valueToString(value: unknown): string | undefined {
  if (typeof value === 'string') {
    return value;
  }
  if (typeof value === 'number' || typeof value === 'boolean') {
    return String(value);
  }
  return undefined;
}

function normalizeSemanticType(value: string): string {
  const normalized = value.trim();
  if (!normalized) {
    return 'schema:Thing';
  }
  if (
    normalized.startsWith('http://') ||
    normalized.startsWith('https://') ||
    normalized.includes(':')
  ) {
    return normalized;
  }
  return `schema:${normalized}`;
}

function deriveSpeculativeName(typeIri: string): string | null {
  const match = typeIri.match(/([^/#]+)$/);
  return match?.[1] ?? null;
}

function resolveSubject(
  declaration: StructuredDataDeclaration,
  routeContext: StructuredDataRouteContext,
  context: StructuredDataContextMap
): StructuredDataNodeRef {
  if (declaration.sdId) {
    const iri = expandIri(declaration.sdId, context, routeContext.baseIri);
    if (iri.startsWith('_:')) {
      return { kind: 'blank', id: iri.slice(2) };
    }
    return { kind: 'iri', iri };
  }

  if (declaration.sdRef) {
    const iri = expandIri(declaration.sdRef, context, routeContext.baseIri);
    if (iri.startsWith('_:')) {
      return { kind: 'blank', id: iri.slice(2) };
    }
    return { kind: 'iri', iri };
  }

  const route = routeContext.routeId ?? '/';
  const line = declaration.sourceSpan?.startLine ?? 0;
  const column = declaration.sourceSpan?.startColumn ?? 0;
  return { kind: 'blank', id: `route-${normalizeNodeId(route)}-${line}-${column}` };
}

function sourceIndexToSpan(
  source: string,
  index: number,
  file: string
): StructuredDataSourceSpan {
  const before = source.slice(0, index);
  const line = before.split('\n').length;
  const lastNewline = before.lastIndexOf('\n');
  const col = index - (lastNewline + 1) + 1;
  return {
    file,
    startLine: line,
    startColumn: col,
    endLine: line,
    endColumn: col + 1,
  };
}

function parseAttributes(tag: string): Record<string, string> {
  const attrs: Record<string, string> = {};
  const attrRegex =
    /\b([A-Za-z_:][A-Za-z0-9_:.-]*)\s*=\s*(?:"([^"]*)"|'([^']*)'|\{\s*"([^"]*)"\s*\}|\{\s*'([^']*)'\s*\})/g;

  for (const match of tag.matchAll(attrRegex)) {
    const key = match[1];
    if (key === undefined) continue;
    const value =
      match[2] ?? match[3] ?? match[4] ?? match[5] ?? '';
    attrs[key] = value;
  }

  return attrs;
}

function parseContextMap(
  raw: string
): { context: StructuredDataContextMap } | { error: string } {
  try {
    const parsed = JSON.parse(raw) as unknown;
    if (!parsed || typeof parsed !== 'object' || Array.isArray(parsed)) {
      return { error: 'context must be a JSON object' };
    }
    const context: StructuredDataContextMap = {};
    for (const [key, value] of Object.entries(parsed as Record<string, unknown>)) {
      if (typeof value !== 'string') {
        return { error: `context prefix '${key}' must map to a string IRI` };
      }
      context[key] = value;
    }
    return { context };
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    return { error: message };
  }
}

function normalizeNodeId(value: string): string {
  return value.replace(/[^a-zA-Z0-9_]/g, '_');
}

function nodeKey(node: StructuredDataNodeRef): string {
  return node.kind === 'iri' ? node.iri : `_:${node.id}`;
}

function serializeNodeRef(node: StructuredDataNodeRef): string {
  return node.kind === 'iri' ? node.iri : `_:${node.id}`;
}

function expandIri(
  raw: string,
  context: StructuredDataContextMap,
  baseIri?: string
): string {
  const value = raw.trim();
  if (value.length === 0) {
    return value;
  }

  if (value.startsWith('_:')) {
    return value;
  }

  if (/^[A-Za-z][A-Za-z0-9+.-]*:\/\//.test(value)) {
    return value;
  }

  if (value.startsWith('#')) {
    if (baseIri) {
      return `${baseIri.replace(/\/$/, '')}${value}`;
    }
    return `_:${value.slice(1)}`;
  }

  const colonIndex = value.indexOf(':');
  if (colonIndex > 0) {
    const prefix = value.slice(0, colonIndex);
    const suffix = value.slice(colonIndex + 1);
    const iriBase = context[prefix];
    if (iriBase) {
      return `${iriBase}${suffix}`;
    }
    if (/^[A-Za-z][A-Za-z0-9+.-]*$/.test(prefix)) {
      return value;
    }
  }

  if (context.schema) {
    return `${context.schema}${value}`;
  }

  return value;
}

function compareDeclarationOrder(
  a: StructuredDataDeclaration,
  b: StructuredDataDeclaration
): number {
  const fa = a.sourceSpan?.file ?? '';
  const fb = b.sourceSpan?.file ?? '';
  if (fa !== fb) {
    return fa.localeCompare(fb);
  }
  const la = a.sourceSpan?.startLine ?? 0;
  const lb = b.sourceSpan?.startLine ?? 0;
  if (la !== lb) {
    return la - lb;
  }
  const ca = a.sourceSpan?.startColumn ?? 0;
  const cb = b.sourceSpan?.startColumn ?? 0;
  return ca - cb;
}

function vocabularyOfIri(iri: string): string | null {
  const hash = iri.lastIndexOf('#');
  if (hash !== -1) {
    return iri.slice(0, hash + 1);
  }
  const slash = iri.lastIndexOf('/');
  if (slash !== -1) {
    return iri.slice(0, slash + 1);
  }
  return null;
}

function literalToJsonLd(
  literal: StructuredDataLiteral
): string | Record<string, string> {
  if (!literal.language) {
    return literal.value;
  }
  return {
    '@value': literal.value,
    '@language': literal.language,
  };
}

function pushJsonLdValue(
  node: Record<string, unknown>,
  key: string,
  value: unknown
): void {
  const current = node[key];
  if (current === undefined) {
    node[key] = value;
    return;
  }
  if (Array.isArray(current)) {
    current.push(value);
    return;
  }
  node[key] = [current, value];
}

function compactIri(iri: string, context: StructuredDataContextMap): string {
  for (const [prefix, value] of Object.entries(context)) {
    if (iri.startsWith(value)) {
      return `${prefix}:${iri.slice(value.length)}`;
    }
  }
  return iri;
}

function groupTriplesBySubject(
  triples: StructuredDataTriple[]
): Map<string, StructuredDataTriple[]> {
  const map = new Map<string, StructuredDataTriple[]>();
  for (const triple of triples) {
    const key = serializeNodeRef(triple.subject);
    const existing = map.get(key) ?? [];
    existing.push(triple);
    map.set(key, existing);
  }
  return map;
}

function toMicrodataType(iri: string): string | null {
  if (!iri.startsWith(SCHEMA_BASE)) {
    return null;
  }
  return iri;
}

function toMicrodataProp(iri: string): string | null {
  if (!iri.startsWith(SCHEMA_BASE)) {
    return null;
  }
  return iri.slice(SCHEMA_BASE.length);
}

function escapeAttr(value: string): string {
  return value
    .replace(/&/g, '&amp;')
    .replace(/"/g, '&quot;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;');
}
