// @a0n/.ignored_gnosis/src/study-graph-snapshot.ts
/**
 * Versioned qualitative / mixed-methods study graph export (structural contract).
 * Canonical semantic graph slice for deterministic validation — not interpretive truth.
 *
 * See docs/STUDY_GRAPH_SNAPSHOT.md.
 */

import type { VerifyResult } from './verify-bundle';

export type StudyMode = 'confirmatory' | 'exploratory' | 'mixed';

export type StudyNodeKind =
  | 'code'
  | 'hypothesis'
  | 'excerpt'
  | 'instrument'
  | 'memo'
  | 'theme';

export type StudyEdgeProvenance = 'human' | 'agent' | 'import';

export interface StudyNodeV1 {
  readonly id: string;
  readonly kind: StudyNodeKind;
  readonly label?: string;
}

export interface StudyEdgeV1 {
  readonly u: string;
  readonly v: string;
  readonly kind?: string;
  readonly provenance?: StudyEdgeProvenance;
}

export interface StudyGraphSnapshotV1 {
  readonly schemaVersion: '1.0.0';
  readonly study_id: string;
  readonly title: string;
  readonly exported_at: string;
  readonly study_mode: StudyMode;
  readonly preregistration_ref?: string;
  readonly plain_language_summary?: string;
  readonly nodes: readonly StudyNodeV1[];
  readonly edges: readonly StudyEdgeV1[];
}

const NODE_KINDS: ReadonlySet<StudyNodeKind> = new Set([
  'code',
  'hypothesis',
  'excerpt',
  'instrument',
  'memo',
  'theme',
]);

const EDGE_PROVENANCE: ReadonlySet<StudyEdgeProvenance> = new Set([
  'human',
  'agent',
  'import',
]);

/**
 * Handles the gnosis validate Study Graph Snapshot workflow.
 */
export function validateStudyGraphSnapshot(x: unknown): VerifyResult {
  if (x === null || typeof x !== 'object') {
    return { ok: false, reason: 'not_object' };
  }
  const o = x as Record<string, unknown>;
  if (o.schemaVersion !== '1.0.0') {
    return { ok: false, reason: 'schema_version' };
  }
  if (typeof o.study_id !== 'string' || o.study_id.length < 2) {
    return { ok: false, reason: 'study_id' };
  }
  if (typeof o.title !== 'string' || o.title.length < 1) {
    return { ok: false, reason: 'title' };
  }
  if (typeof o.exported_at !== 'string' || Number.isNaN(Date.parse(o.exported_at))) {
    return { ok: false, reason: 'exported_at' };
  }
  if (
    o.study_mode !== 'confirmatory' &&
    o.study_mode !== 'exploratory' &&
    o.study_mode !== 'mixed'
  ) {
    return { ok: false, reason: 'study_mode' };
  }
  if (
    o.preregistration_ref !== undefined &&
    typeof o.preregistration_ref !== 'string'
  ) {
    return { ok: false, reason: 'preregistration_ref_type' };
  }
  if (
    o.plain_language_summary !== undefined &&
    typeof o.plain_language_summary !== 'string'
  ) {
    return { ok: false, reason: 'plain_language_summary_type' };
  }
  const nodes = o.nodes;
  if (!Array.isArray(nodes) || nodes.length < 1) {
    return { ok: false, reason: 'nodes' };
  }
  const ids = new Set<string>();
  for (let i = 0; i < nodes.length; i++) {
    const n = nodes[i];
    if (n === null || typeof n !== 'object') {
      return { ok: false, reason: `node_${i}` };
    }
    const nr = n as Record<string, unknown>;
    if (typeof nr.id !== 'string' || nr.id.length < 1) {
      return { ok: false, reason: `node_${i}_id` };
    }
    if (ids.has(nr.id)) {
      return { ok: false, reason: 'duplicate_node_id' };
    }
    ids.add(nr.id);
    if (typeof nr.kind !== 'string' || !NODE_KINDS.has(nr.kind as StudyNodeKind)) {
      return { ok: false, reason: `node_${i}_kind` };
    }
    if (nr.label !== undefined && typeof nr.label !== 'string') {
      return { ok: false, reason: `node_${i}_label` };
    }
  }
  const edges = o.edges;
  if (!Array.isArray(edges)) {
    return { ok: false, reason: 'edges' };
  }
  for (let j = 0; j < edges.length; j++) {
    const e = edges[j];
    if (e === null || typeof e !== 'object') {
      return { ok: false, reason: `edge_${j}` };
    }
    const er = e as Record<string, unknown>;
    if (typeof er.u !== 'string' || !ids.has(er.u)) {
      return { ok: false, reason: `edge_${j}_u` };
    }
    if (typeof er.v !== 'string' || !ids.has(er.v)) {
      return { ok: false, reason: `edge_${j}_v` };
    }
    if (er.kind !== undefined && typeof er.kind !== 'string') {
      return { ok: false, reason: `edge_${j}_kind` };
    }
    if (er.provenance !== undefined) {
      const p = er.provenance;
      if (
        typeof p !== 'string' ||
        !EDGE_PROVENANCE.has(p as StudyEdgeProvenance)
      ) {
        return { ok: false, reason: `edge_${j}_provenance` };
      }
    }
  }
  return { ok: true };
}
