// @a0n/.ignored_gnosis/src/study-publish-gate.ts
/**
 * Deterministic publish gate for study graph snapshots (arena-merge-gate shaped).
 * Pure functions; no I/O. Structural + policy checks only — not ethics/IRB.
 */

import type { StudyGraphSnapshotV1 } from './study-graph-snapshot';
import { validateStudyGraphSnapshot } from './study-graph-snapshot';

export interface StudyPublishGateInput {
  readonly studyGraphSnapshot: unknown;
}

export type StudyPublishGateResult =
  | { ok: true }
  | { ok: false; reasons: readonly string[] };

function isSnapshotV1(x: unknown): x is StudyGraphSnapshotV1 {
  return (
    x !== null &&
    typeof x === 'object' &&
    (x as StudyGraphSnapshotV1).schemaVersion === '1.0.0'
  );
}

/**
 * Handles the gnosis evaluate Study Publish Gate workflow.
 */
export function evaluateStudyPublishGate(
  input: StudyPublishGateInput
): StudyPublishGateResult {
  const v = validateStudyGraphSnapshot(input.studyGraphSnapshot);
  if (v.ok === false) {
    return { ok: false, reasons: [`snapshot_${v.reason}`] };
  }
  if (!isSnapshotV1(input.studyGraphSnapshot)) {
    return { ok: false, reasons: ['snapshot_shape'] };
  }
  const s = input.studyGraphSnapshot;
  const reasons: string[] = [];

  if (s.study_mode === 'confirmatory' || s.study_mode === 'mixed') {
    const pr = s.preregistration_ref?.trim();
    if (pr === undefined || pr.length < 4) {
      reasons.push('confirmatory_or_mixed_requires_preregistration_ref');
    }
  }

  if (s.study_mode === 'confirmatory') {
    const hasHypothesis = s.nodes.some((n) => n.kind === 'hypothesis');
    if (!hasHypothesis) {
      reasons.push('confirmatory_requires_hypothesis_node');
    }
  }

  if (reasons.length > 0) {
    return { ok: false, reasons };
  }
  return { ok: true };
}
