// @a0n/.ignored_gnosis/src/study-email-templates.ts
/**
 * Pure HTML/text/subject builders for study transactional mail.
 * Sending happens in apps/edgework-app (Mailgun HTTP) — see docs/STUDY_PARTICIPANT_DATA_POLICY.md
 */

export interface StudyInviteTemplateInput {
  readonly studyTitle: string;
  readonly studyId: string;
  /** HTTPS URL for consent / instrument entry */
  readonly actionUrl: string;
  readonly consentVersion?: string;
}

export interface StudyReminderTemplateInput {
  readonly studyTitle: string;
  readonly studyId: string;
  readonly actionUrl: string;
}

function escapeHtml(s: string): string {
  return s
    .replace(/&/gu, '&amp;')
    .replace(/</gu, '&lt;')
    .replace(/>/gu, '&gt;')
    .replace(/"/gu, '&quot;');
}

/**
 * Handles the gnosis render Study Invite Email workflow.
 */
export function renderStudyInviteEmail(
  input: StudyInviteTemplateInput
): { subject: string; html: string; text: string } {
  const subject = `Study invitation: ${input.studyTitle}`;
  const cv = input.consentVersion?.trim();
  const safeTitle = escapeHtml(input.studyTitle);
  const safeId = escapeHtml(input.studyId);
  const safeUrl = escapeHtml(input.actionUrl);
  const html = `<!DOCTYPE html>
<html><head><meta charset="utf-8"/><meta name="description" content="Study invitation details and secure participation link."/></head>
<body style="font-family:system-ui,Segoe UI,sans-serif;line-height:1.5;color:#111;max-width:560px;margin:0 auto;padding:24px;">
  <h1 style="font-size:1.25rem;">You are invited</h1>
  <p><strong>${safeTitle}</strong></p>
  <p style="font-size:0.875rem;color:#444;">Study ID: <code>${safeId}</code></p>
  ${cv ? `<p style="font-size:0.875rem;">Consent / protocol version: <strong>${escapeHtml(cv)}</strong></p>` : ''}
  <p><a href="${safeUrl}" style="display:inline-block;padding:10px 18px;background:#0d9488;color:#fff;text-decoration:none;border-radius:6px;">Open study link</a></p>
  <p style="font-size:0.8rem;color:#666;">If you did not expect this message, you can ignore it.</p>
</body></html>`;
  const text = [
    `You are invited to participate: ${input.studyTitle}`,
    `Study ID: ${input.studyId}`,
    cv ? `Consent version: ${cv}` : '',
    `Link: ${input.actionUrl}`,
  ]
    .filter(Boolean)
    .join('\n');
  return { subject, html, text };
}

/**
 * Handles the gnosis render Study Reminder Email workflow.
 */
export function renderStudyReminderEmail(
  input: StudyReminderTemplateInput
): { subject: string; html: string; text: string } {
  const subject = `Reminder: ${input.studyTitle}`;
  const safeTitle = escapeHtml(input.studyTitle);
  const safeId = escapeHtml(input.studyId);
  const safeUrl = escapeHtml(input.actionUrl);
  const html = `<!DOCTYPE html>
<html><head><meta charset="utf-8"/><meta name="description" content="Study reminder with secure continuation link."/></head>
<body style="font-family:system-ui,Segoe UI,sans-serif;line-height:1.5;color:#111;max-width:560px;margin:0 auto;padding:24px;">
  <h2 style="font-size:1.25rem;">Reminder</h2>
  <p><strong>${safeTitle}</strong></p>
  <p style="font-size:0.875rem;color:#444;">Study ID: <code>${safeId}</code></p>
  <p><a href="${safeUrl}" style="display:inline-block;padding:10px 18px;background:#0d9488;color:#fff;text-decoration:none;border-radius:6px;">Continue</a></p>
</body></html>`;
  const text = [
    `Reminder: ${input.studyTitle}`,
    `Study ID: ${input.studyId}`,
    `Link: ${input.actionUrl}`,
  ].join('\n');
  return { subject, html, text };
}
