// @a0n/.ignored_gnosis/src/symmetry-conservation.ts
/**
 * symmetry-conservation.ts -- Noether's theorem for topologies
 *
 * The identification:
 *   - Topology symmetry = relabeling invariance
 *   - Conserved quantity = beta-1 (first Betti number)
 *   - Time-translation symmetry = PROCESS invariance
 *   - Permutation symmetry = FORK commutativity
 *   - Gauge symmetry = VoidBoundary relabeling
 *
 * Noether's theorem: every continuous symmetry of a topology
 * corresponds to a conserved quantity of its execution.
 */

import {
  type VoidBoundary,
  complementDistribution,
  shannonEntropy,
} from './void';

// ============================================================================
// Symmetry detection
// ============================================================================

export type SymmetryKind =
  | 'permutation' // Relabeling nodes preserves topology
  | 'time-translation' // Shifting all depths preserves dynamics
  | 'scale' // Multiplying all counts preserves complement distribution shape
  | 'reflection' // Reversing edge direction preserves beta-1
  | 'gauge'; // Relabeling VoidBoundary dimensions

export interface Symmetry {
  kind: SymmetryKind;
  /** The permutation/transformation */
  transform: number[];
  /** Description */
  description: string;
}

/**
 * Check if a permutation of node indices preserves the adjacency structure.
 */
export function isAutomorphism(
  edges: [number, number][],
  permutation: number[]
): boolean {
  const permutedEdges = new Set(
    edges.map(([s, t]) => `${permutation[s]}-${permutation[t]}`)
  );
  const originalEdges = new Set(edges.map(([s, t]) => `${s}-${t}`));
  if (permutedEdges.size !== originalEdges.size) return false;
  for (const e of originalEdges) {
    if (!permutedEdges.has(e)) return false;
  }
  return true;
}

/**
 * Find all permutation symmetries (graph automorphisms).
 * For small graphs only (exponential in general).
 */
export function findAutomorphisms(
  n: number,
  edges: [number, number][]
): number[][] {
  if (n > 8) return []; // Only tractable for small graphs
  const automorphisms: number[][] = [];

  function permute(arr: number[], l: number): void {
    if (l === n) {
      if (isAutomorphism(edges, arr)) {
        automorphisms.push([...arr]);
      }
      return;
    }
    for (let i = l; i < n; i++) {
      [arr[l], arr[i]] = [arr[i], arr[l]];
      permute(arr, l + 1);
      [arr[l], arr[i]] = [arr[i], arr[l]];
    }
  }

  permute(
    Array.from({ length: n }, (_, i) => i),
    0
  );
  return automorphisms;
}

/**
 * Check if a VoidBoundary has scale symmetry:
 * multiplying all counts by a constant preserves the complement distribution.
 */
export function hasScaleSymmetry(
  boundary: VoidBoundary,
  eta: number,
  tolerance: number = 1e-10
): boolean {
  const dist1 = complementDistribution(boundary, eta);
  // Scale counts by 2
  const scaled = {
    counts: boundary.counts.map((c) => c * 2),
    totalEntries: boundary.totalEntries * 2,
  };
  const dist2 = complementDistribution(scaled, eta);
  return dist1.every((p, i) => Math.abs(p - dist2[i]) < tolerance);
}

// ============================================================================
// Conserved quantities
// ============================================================================

export interface ConservedQuantity {
  name: string;
  symmetry: SymmetryKind;
  value: number;
  /** Noether charge: the quantity that is conserved */
  description: string;
}

/**
 * Beta-1 conservation: fork/fold balance.
 * Conserved under PROCESS (time-translation) symmetry.
 */
export function beta1Conservation(
  forkDims: number,
  foldDims: number
): ConservedQuantity {
  return {
    name: 'beta-1',
    symmetry: 'time-translation',
    value: forkDims - foldDims,
    description:
      'First Betti number: fork creates, fold destroys. Net = 0 at boundary.',
  };
}

/**
 * Entropy conservation under permutation symmetry.
 * If the topology has a permutation symmetry, the Shannon entropy
 * of the complement distribution is invariant under that permutation.
 */
export function entropyConservation(
  boundary: VoidBoundary,
  eta: number
): ConservedQuantity {
  const dist = complementDistribution(boundary, eta);
  return {
    name: 'shannon-entropy',
    symmetry: 'permutation',
    value: shannonEntropy(dist),
    description: 'Shannon entropy is invariant under dimension permutation.',
  };
}

/**
 * Total void conservation: void is never destroyed, only created.
 * This is the "charge conservation" of void walking -- the total
 * void count is monotonically non-decreasing.
 */
export function voidChargeConservation(
  boundary: VoidBoundary
): ConservedQuantity {
  return {
    name: 'void-charge',
    symmetry: 'time-translation',
    value: boundary.totalEntries,
    description: 'Total void entries never decrease (second law).',
  };
}

/**
 * Verify Noether's theorem: for each symmetry, check the conserved quantity
 * is preserved under the symmetry transformation.
 */
export function verifyNoether(
  boundary: VoidBoundary,
  permutation: number[],
  eta: number
): { conserved: boolean; entropyBefore: number; entropyAfter: number } {
  const distBefore = complementDistribution(boundary, eta);
  const entropyBefore = shannonEntropy(distBefore);

  // Apply permutation to boundary
  const permuted: VoidBoundary = {
    counts: permutation.map((i) => boundary.counts[i]),
    totalEntries: boundary.totalEntries,
  };
  const distAfter = complementDistribution(permuted, eta);
  const entropyAfter = shannonEntropy(distAfter);

  return {
    conserved: Math.abs(entropyBefore - entropyAfter) < 1e-10,
    entropyBefore,
    entropyAfter,
  };
}
