// @a0n/.ignored_gnosis/src/temporal-prism.ts
/**
 * Temporal Prism — keep a message AWAY from its target for a fixed duration.
 *
 * Standard routing minimizes latency to target. The temporal prism does
 * the opposite: it routes a message on a path whose total delay equals
 * a specified duration, keeping the message "in flight" until its
 * deadline. Like light refracting through glass, the message bends
 * along the path that delays it the right amount.
 *
 * The clock topology provides exact edge weights (propagation delays
 * between AeonPIDs). The prism finds a cycle or path whose sum equals
 * the target delay, preferring hops through PIDs in active solar phase
 * to ensure the message always lives in a warm node during transit.
 *
 * Applications:
 *   - Time-locked payloads (arrive at specific pLo)
 *   - Delayed consensus (circulate until quorum)
 *   - Anti-thundering-herd (naturally stagger coordinated events)
 *   - Chain-of-custody witnessing (prove message traversed a path)
 *   - Rate-limit evasion-by-delay (arrive after the window expires)
 *
 * @packageDocumentation
 */

import type { DistributedClockMesh, ProcessClock } from './clock-distributed';
import type { SolarObserver } from './math';
import { solarPosition } from './math';
import { classifyPhaseDirection, phaseScore } from './solar-scheduler';

// ─── Types ──────────────────────────────────────────────────────────

/**
 * A single hop in the prism path.
 */
export interface PrismHop {
  readonly fromPidId: string;
  readonly toPidId: string;
  readonly fromLabel: string;
  readonly toLabel: string;
  readonly propagationDelayMs: number;
  readonly departureEpochMs: number;
  readonly arrivalEpochMs: number;
  /** The solar phase at the destination when the message arrives. */
  readonly arrivalPhase: 'rising' | 'peak' | 'falling' | 'night';
  /** Phase attractiveness of the destination (higher = warmer target). */
  readonly phaseScore: number;
}

/**
 * A full prism path: the message's journey from origin to eventual target.
 */
export interface PrismPath {
  readonly source: 'GnosisTemporalPrism';
  readonly originPidId: string;
  readonly targetPidId: string;
  readonly startEpochMs: number;
  readonly deliverAtEpochMs: number;
  readonly hops: readonly PrismHop[];
  readonly totalDelayMs: number;
  readonly actualArrivalEpochMs: number;
  readonly deliveryErrorMs: number; // |actualArrival - deliverAt|
  /** Average phase score across all hops -- higher = warmer path. */
  readonly averagePhaseScore: number;
  /** Whether the path met the target delivery time within tolerance. */
  readonly onTime: boolean;
}

// ─── Prism Options ──────────────────────────────────────────────────

export interface PrismOptions {
  /** Tolerance for delivery time matching, in milliseconds. Default: 5% of totalDelay. */
  readonly toleranceMs?: number;
  /** Maximum number of hops to consider. Default: 16. */
  readonly maxHops?: number;
  /** Penalty for hops through night-phase nodes (0-1). Default: 0.3. */
  readonly nightPenalty?: number;
  /** Random seed for path exploration (deterministic if set). */
  readonly seed?: number;
}

// ─── Core Algorithm ─────────────────────────────────────────────────

/**
 * Build a temporal prism path.
 *
 * Finds a sequence of intermediate PIDs such that the total propagation
 * delay from origin through the chain to target equals deliveryDelayMs.
 *
 * Algorithm: depth-first greedy search with delay budgeting.
 * At each hop, pick the intermediate PID that:
 *   1. Brings remaining budget closer to being absorbable by direct path
 *   2. Is in an active/rising solar phase (warm, available)
 *   3. Has not been visited (avoid cycles unless budget requires revisit)
 *
 * Returns null if no valid path found within maxHops.
 */
export function buildPrismPath(
  mesh: DistributedClockMesh,
  originPidId: string,
  targetPidId: string,
  deliveryDelayMs: number,
  options: PrismOptions = {}
): PrismPath | null {
  const {
    toleranceMs = Math.max(10, deliveryDelayMs * 0.05),
    maxHops = 16,
    nightPenalty = 0.3,
  } = options;

  const origin = mesh.clocks.find((c: ProcessClock) => c.id === originPidId);
  const target = mesh.clocks.find((c: ProcessClock) => c.id === targetPidId);
  if (!origin || !target) return null;

  const startEpochMs = Date.now();
  const deliverAtEpochMs = startEpochMs + deliveryDelayMs;

  // Precompute edge map: delay[fromId][toId] = propagationMs
  const edgeMap = new Map<string, Map<string, number>>();
  for (const edge of mesh.edges) {
    if (!edgeMap.has(edge.fromId)) edgeMap.set(edge.fromId, new Map());
    if (!edgeMap.has(edge.toId)) edgeMap.set(edge.toId, new Map());
    edgeMap.get(edge.fromId)!.set(edge.toId, edge.distance.propagationDelayMs);
    edgeMap.get(edge.toId)!.set(edge.fromId, edge.distance.propagationDelayMs);
  }

  const clockById = new Map<string, ProcessClock>(
    mesh.clocks.map((c: ProcessClock) => [c.id, c] as const)
  );

  // DFS with budget tracking
  const visited = new Set<string>([originPidId]);
  const path: PrismHop[] = [];

  function dfs(currentId: string, remainingBudgetMs: number, currentEpochMs: number): boolean {
    // Direct hop to target possible?
    const directDelay = edgeMap.get(currentId)?.get(targetPidId) ?? Infinity;
    if (Math.abs(directDelay - remainingBudgetMs) <= toleranceMs) {
      // Complete the path
      const hop = makeHop(currentId, targetPidId, directDelay, currentEpochMs, clockById);
      if (hop) path.push(hop);
      return true;
    }

    if (path.length >= maxHops - 1) return false;

    // Candidate next hops: all reachable PIDs that are not target (we save target for last)
    const neighbors = edgeMap.get(currentId);
    if (!neighbors) return false;

    const candidates: Array<{ id: string; delay: number; score: number }> = [];
    for (const [nextId, delay] of neighbors) {
      if (nextId === targetPidId) continue; // reserve target for final hop
      if (visited.has(nextId)) continue;
      if (delay > remainingBudgetMs) continue; // already past budget
      // After this hop, we need a path from nextId to target summing to (remainingBudget - delay)
      const nextRemaining = remainingBudgetMs - delay;
      if (nextRemaining < 0) continue;
      // Heuristic: prefer hops whose remaining budget is ~= delay to target
      const toTargetFromNext = edgeMap.get(nextId)?.get(targetPidId) ?? Infinity;
      const budgetGap = Math.abs(toTargetFromNext - nextRemaining);
      // Phase score at next hop's arrival time
      const nextClock = clockById.get(nextId);
      let pScore = 0.5;
      if (nextClock) {
        const arrivalMs = currentEpochMs + delay;
        const pos = solarPosition(
          { latitudeDeg: nextClock.location.lat, longitudeDeg: nextClock.location.lon },
          arrivalMs
        );
        // Rough slope estimate: difference with 10 min earlier
        const prior = solarPosition(
          { latitudeDeg: nextClock.location.lat, longitudeDeg: nextClock.location.lon },
          arrivalMs - 600_000
        );
        const slope = (pos.altitudeDeg - prior.altitudeDeg) * 6; // per hour
        const direction = classifyPhaseDirection(pos.altitudeDeg, slope);
        pScore = phaseScore(direction);
        if (direction === 'night') pScore *= (1 - nightPenalty);
      }
      // Combined candidate score: smaller budget gap and higher phase score = better
      const candidateScore = pScore / (1 + budgetGap / 1000);
      candidates.push({ id: nextId, delay, score: candidateScore });
    }

    candidates.sort((a, b) => b.score - a.score);

    for (const cand of candidates) {
      const hop = makeHop(currentId, cand.id, cand.delay, currentEpochMs, clockById);
      if (!hop) continue;
      path.push(hop);
      visited.add(cand.id);
      if (dfs(cand.id, remainingBudgetMs - cand.delay, hop.arrivalEpochMs)) {
        return true;
      }
      path.pop();
      visited.delete(cand.id);
    }

    return false;
  }

  const found = dfs(originPidId, deliveryDelayMs, startEpochMs);
  if (!found) return null;

  const totalDelayMs = path.reduce((sum, hop) => sum + hop.propagationDelayMs, 0);
  const actualArrivalMs = startEpochMs + totalDelayMs;
  const deliveryErrorMs = Math.abs(actualArrivalMs - deliverAtEpochMs);
  const avgPhase = path.length > 0
    ? path.reduce((sum, hop) => sum + hop.phaseScore, 0) / path.length
    : 0;

  return {
    source: 'GnosisTemporalPrism',
    originPidId,
    targetPidId,
    startEpochMs,
    deliverAtEpochMs,
    hops: path,
    totalDelayMs,
    actualArrivalEpochMs: actualArrivalMs,
    deliveryErrorMs,
    averagePhaseScore: avgPhase,
    onTime: deliveryErrorMs <= toleranceMs,
  };
}

// ─── Hop Construction Helper ────────────────────────────────────────

function makeHop(
  fromId: string,
  toId: string,
  propagationMs: number,
  departureEpochMs: number,
  clockById: ReadonlyMap<string, ProcessClock>
): PrismHop | null {
  const from = clockById.get(fromId);
  const to = clockById.get(toId);
  if (!from || !to) return null;

  const arrivalEpochMs = departureEpochMs + propagationMs;

  // Phase at arrival
  const pos = solarPosition(
    { latitudeDeg: to.location.lat, longitudeDeg: to.location.lon },
    arrivalEpochMs
  );
  const prior = solarPosition(
    { latitudeDeg: to.location.lat, longitudeDeg: to.location.lon },
    arrivalEpochMs - 600_000
  );
  const slope = (pos.altitudeDeg - prior.altitudeDeg) * 6;
  const direction = classifyPhaseDirection(pos.altitudeDeg, slope);

  return {
    fromPidId: fromId,
    toPidId: toId,
    fromLabel: from.label,
    toLabel: to.label,
    propagationDelayMs: propagationMs,
    departureEpochMs,
    arrivalEpochMs,
    arrivalPhase: direction,
    phaseScore: phaseScore(direction),
  };
}

// ─── Analysis ───────────────────────────────────────────────────────

/**
 * Summary stats about a prism path.
 */
export interface PrismAnalysis {
  readonly source: 'GnosisPrismAnalysis';
  readonly hopCount: number;
  readonly totalDelayMs: number;
  readonly deliveryErrorMs: number;
  readonly onTime: boolean;
  readonly uniquePidsTraversed: number;
  readonly geographicSpreadDeg: number; // max - min longitude across hops
  readonly nightHopFraction: number;    // fraction of hops arriving in night phase
  readonly risingHopFraction: number;   // fraction arriving in rising phase
}

/**
 * Analyze a prism path for visualization and auditing.
 */
export function analyzePrism(
  path: PrismPath,
  mesh: DistributedClockMesh
): PrismAnalysis {
  const clockById = new Map<string, ProcessClock>(
    mesh.clocks.map((c: ProcessClock) => [c.id, c] as const)
  );
  const pidsTraversed = new Set<string>([path.originPidId]);
  for (const hop of path.hops) {
    pidsTraversed.add(hop.toPidId);
  }

  const lons: number[] = [];
  for (const id of pidsTraversed) {
    const clock = clockById.get(id);
    if (clock) lons.push(clock.location.lon);
  }
  const geographicSpreadDeg = lons.length > 0
    ? Math.max(...lons) - Math.min(...lons)
    : 0;

  const nightCount = path.hops.filter(h => h.arrivalPhase === 'night').length;
  const risingCount = path.hops.filter(h => h.arrivalPhase === 'rising').length;

  return {
    source: 'GnosisPrismAnalysis',
    hopCount: path.hops.length,
    totalDelayMs: path.totalDelayMs,
    deliveryErrorMs: path.deliveryErrorMs,
    onTime: path.onTime,
    uniquePidsTraversed: pidsTraversed.size,
    geographicSpreadDeg,
    nightHopFraction: path.hops.length > 0 ? nightCount / path.hops.length : 0,
    risingHopFraction: path.hops.length > 0 ? risingCount / path.hops.length : 0,
  };
}

// ─── Slowest-Possible-Path ──────────────────────────────────────────

export interface SlowestPrismOptions {
  /** Maximum number of hops allowed. Default: all nodes minus target minus origin. */
  readonly maxHops?: number;
  /** Upper ceiling on total delay (ms). Default: no ceiling. */
  readonly ceilingMs?: number;
  /** If true, visit every reachable node exactly once before delivering. */
  readonly witnessAll?: boolean;
}

/**
 * Build the slowest possible prism path from origin to target.
 *
 * Where `buildPrismPath` matches a specific delay budget, this finds the
 * path that MAXIMIZES total in-flight time. It visits as many distinct
 * intermediate PIDs as possible before delivering, each hop accumulating
 * more delay.
 *
 * Use cases:
 *   - Load testing / worst-case latency probing
 *   - Evidence accumulation (each hop signs the message)
 *   - Maximum-entropy routing (touch every PID before delivery)
 *   - Anti-brute-force delays (frustrate timing attacks)
 *   - Deliberate cooling periods
 *
 * Longest-path is NP-hard in general, but with the clock mesh's structure
 * (fully connected, metric edge weights) greedy "visit nearest unvisited"
 * gives a reasonable approximation. If `witnessAll` is true, the algorithm
 * uses a nearest-neighbor tour through all PIDs (a TSP heuristic) before
 * the final hop to target. Otherwise it uses a greedy "prefer the longest
 * unvisited edge" expansion until the ceiling is reached or all nodes visited.
 */
export function buildSlowestPrismPath(
  mesh: DistributedClockMesh,
  originPidId: string,
  targetPidId: string,
  options: SlowestPrismOptions = {}
): PrismPath | null {
  const origin = mesh.clocks.find((c: ProcessClock) => c.id === originPidId);
  const target = mesh.clocks.find((c: ProcessClock) => c.id === targetPidId);
  if (!origin || !target) return null;

  const {
    maxHops = mesh.clocks.length + 1,
    ceilingMs = Infinity,
    witnessAll = false,
  } = options;

  const startEpochMs = Date.now();

  // Precompute edge weights
  const edgeMap = new Map<string, Map<string, number>>();
  for (const edge of mesh.edges) {
    if (!edgeMap.has(edge.fromId)) edgeMap.set(edge.fromId, new Map());
    if (!edgeMap.has(edge.toId)) edgeMap.set(edge.toId, new Map());
    edgeMap.get(edge.fromId)!.set(edge.toId, edge.distance.propagationDelayMs);
    edgeMap.get(edge.toId)!.set(edge.fromId, edge.distance.propagationDelayMs);
  }

  const clockById = new Map<string, ProcessClock>(
    mesh.clocks.map((c: ProcessClock) => [c.id, c] as const)
  );

  // Greedy: at each step, pick the longest edge to an unvisited non-target node.
  // Stop when all non-target nodes visited, maxHops-1 reached, or ceiling would be exceeded by the next hop + target hop.
  const visited = new Set<string>([originPidId]);
  const path: PrismHop[] = [];
  let currentId = originPidId;
  let currentEpochMs = startEpochMs;
  let accumulatedMs = 0;

  while (true) {
    // How much room remains before ceiling, accounting for the final hop to target?
    const directToTarget = edgeMap.get(currentId)?.get(targetPidId) ?? Infinity;
    const remainingBeforeCeiling = ceilingMs - accumulatedMs;
    if (!Number.isFinite(directToTarget)) break; // no way to reach target
    if (path.length >= maxHops - 1) break; // only target hop left
    if (witnessAll && visited.size >= mesh.clocks.length) break; // visited everyone

    // Candidates: unvisited, non-target, with propagation such that the eventual
    // target hop still fits under the ceiling.
    const neighbors = edgeMap.get(currentId);
    if (!neighbors) break;

    const candidates: Array<{ id: string; delay: number }> = [];
    for (const [nextId, delay] of neighbors) {
      if (visited.has(nextId)) continue;
      if (nextId === targetPidId) continue;
      // Will this hop + eventual target hop still fit?
      const minFutureDelay = delay;
      const directFromNext = edgeMap.get(nextId)?.get(targetPidId) ?? Infinity;
      if (!Number.isFinite(directFromNext)) continue;
      if (accumulatedMs + minFutureDelay + directFromNext > ceilingMs) continue;
      candidates.push({ id: nextId, delay });
    }

    if (candidates.length === 0) break;

    // Pick the LONGEST edge (greedy maximization)
    candidates.sort((a, b) => b.delay - a.delay);
    const chosen = candidates[0];

    const hop = makeHop(currentId, chosen.id, chosen.delay, currentEpochMs, clockById);
    if (!hop) break;
    path.push(hop);
    visited.add(chosen.id);
    currentId = chosen.id;
    currentEpochMs = hop.arrivalEpochMs;
    accumulatedMs += chosen.delay;
  }

  // Final hop to target
  const finalDelay = edgeMap.get(currentId)?.get(targetPidId);
  if (finalDelay === undefined) return null;
  if (accumulatedMs + finalDelay > ceilingMs) return null;

  const finalHop = makeHop(currentId, targetPidId, finalDelay, currentEpochMs, clockById);
  if (!finalHop) return null;
  path.push(finalHop);

  const totalDelayMs = path.reduce((sum, hop) => sum + hop.propagationDelayMs, 0);
  const actualArrivalMs = startEpochMs + totalDelayMs;
  const avgPhase = path.length > 0
    ? path.reduce((sum, hop) => sum + hop.phaseScore, 0) / path.length
    : 0;

  return {
    source: 'GnosisTemporalPrism',
    originPidId,
    targetPidId,
    startEpochMs,
    deliverAtEpochMs: actualArrivalMs, // "deliver as late as possible" -- no pre-set target
    hops: path,
    totalDelayMs,
    actualArrivalEpochMs: actualArrivalMs,
    deliveryErrorMs: 0, // slowest path has no "target" to miss
    averagePhaseScore: avgPhase,
    onTime: true, // always "on time" when the time IS as-late-as-possible
  };
}

/**
 * The theoretical maximum delay achievable for a prism from origin to target.
 *
 * Walks the mesh edges and sums the weights of a nearest-neighbor tour
 * visiting every PID exactly once, plus the final hop to target. This is
 * an upper bound for `buildSlowestPrismPath` -- the actual path may be
 * shorter if the ceiling or maxHops cut it off.
 */
export function maxPossibleDelayMs(
  mesh: DistributedClockMesh,
  originPidId: string,
  targetPidId: string
): number {
  const path = buildSlowestPrismPath(mesh, originPidId, targetPidId, {
    witnessAll: true,
  });
  return path?.totalDelayMs ?? 0;
}
