// @a0n/.ignored_gnosis/src/thermodynamic.ts
/**
 * thermodynamic.ts -- Landauer erasure + Boltzmann distribution as void walking
 *
 * The identification:
 *   - VENT = irreversible erasure = kT ln 2 per bit (Landauer's principle)
 *   - complementDistribution = Boltzmann distribution (e^{-βE} / Z)
 *   - void monotonically increases = second law of thermodynamics
 *   - eta = inverse temperature β = 1/kT
 *   - VoidBoundary.totalEntries = total dissipated work
 *   - fork creates entropy (increases phase space)
 *   - fold reduces entropy (collapses microstates)
 *   - the free energy F = E - TS is the walker's score minus entropy times temperature
 *
 * The formal ledger entry: every void walk is a thermodynamic process.
 * The complement distribution formalizes the canonical ensemble. The boundary
 * formalizes the energy landscape. The walker maps to a heat engine.
 */

import {
  type VoidBoundary,
  type Walker,
  type Measurement,
  createVoidBoundary,
  updateVoidBoundary,
  complementDistribution,
  shannonEntropy,
  measure,
  createWalker,
  c0Choose,
  c0Update,
  c1Measure,
} from './void';

// ============================================================================
// Constants -- connect void walking to physics
// ============================================================================

/** Boltzmann constant in natural units (nats per unit temperature) */
export const K_BOLTZMANN = 1.0;

/** Landauer limit: minimum energy to erase one bit at temperature T */
export function landauerLimit(temperature: number): number {
  return K_BOLTZMANN * temperature * Math.LN2;
}

// ============================================================================
// Thermodynamic state -- void boundary as energy landscape
// ============================================================================

export interface ThermodynamicState {
  /** The void boundary formalizes the energy landscape */
  boundary: VoidBoundary;
  /** Inverse temperature β = eta = 1/kT */
  beta: number;
  /** Temperature T = 1/(k β) */
  temperature: number;
  /** Total dissipated work (Landauer cost of all venting) */
  dissipatedWork: number;
  /** Total entropy production */
  entropyProduction: number;
  /** Number of irreversible erasures (VENT operations) */
  erasures: number;
  /** Partition function Z = Σ exp(-β E_i) */
  partitionFunction: number;
  /** Free energy F = -T ln Z */
  freeEnergy: number;
  /** Internal energy U = Σ E_i p_i */
  internalEnergy: number;
  /** Thermodynamic Buffer: distance from saturation boundary (U - erasures) */
  thermodynamicBuffer: number;
  /** Manifold Depth: current expected energy level (U) */
  manifoldDepth: number;
}

/**
 * Create a thermodynamic state from a void boundary.
 * The boundary counts ARE energy levels.
 * eta is inverse temperature β.
 */
export function createThermodynamicState(
  boundary: VoidBoundary,
  beta: number
): ThermodynamicState {
  const temperature = 1 / (K_BOLTZMANN * beta);
  const { partitionFunction, freeEnergy, internalEnergy } =
    computeThermodynamicPotentials(boundary, beta, temperature);

  return {
    boundary,
    beta,
    temperature,
    dissipatedWork: 0,
    entropyProduction: 0,
    erasures: 0,
    partitionFunction,
    freeEnergy,
    internalEnergy,
    thermodynamicBuffer: internalEnergy, // Initially no erasures
    manifoldDepth: internalEnergy,
  };
}

function computeThermodynamicPotentials(
  boundary: VoidBoundary,
  beta: number,
  temperature: number
): { partitionFunction: number; freeEnergy: number; internalEnergy: number } {
  // Z = Σ exp(-β E_i) where E_i = counts[i]
  let Z = 0;
  for (const E of boundary.counts) {
    Z += Math.exp(-beta * E);
  }

  const freeEnergy = -temperature * Math.log(Z);

  // U = Σ E_i p_i = Σ E_i exp(-β E_i) / Z
  let U = 0;
  for (const E of boundary.counts) {
    U += (E * Math.exp(-beta * E)) / Z;
  }

  return { partitionFunction: Z, freeEnergy, internalEnergy: U };
}

/**
 * Refresh thermodynamic potentials after boundary changes.
 */
export function refreshPotentials(state: ThermodynamicState): void {
  const { partitionFunction, freeEnergy, internalEnergy } =
    computeThermodynamicPotentials(
      state.boundary,
      state.beta,
      state.temperature
    );
  state.partitionFunction = partitionFunction;
  state.freeEnergy = freeEnergy;
  state.internalEnergy = internalEnergy;
  state.thermodynamicBuffer = internalEnergy - state.erasures;
  state.manifoldDepth = internalEnergy;
}

// ============================================================================
// The Boltzmann identification
// ============================================================================

/**
 * The complement distribution formalizes the Boltzmann distribution.
 *
 * complementDistribution: w_i = exp(-eta * normalized_count_i) / Σ
 * Boltzmann distribution: p_i = exp(-β E_i) / Z
 *
 * When eta = β and counts = energies, these are the same function.
 * The proof is structural: both are softmax over negative energies.
 */
export function boltzmannDistribution(state: ThermodynamicState): number[] {
  return complementDistribution(state.boundary, state.beta);
}

/**
 * Gibbs entropy S = -k Σ p_i ln p_i
 * This is shannonEntropy in natural units times Boltzmann constant.
 */
export function gibbsEntropy(state: ThermodynamicState): number {
  const dist = boltzmannDistribution(state);
  return K_BOLTZMANN * shannonEntropy(dist);
}

/**
 * Verify the fundamental thermodynamic identity: F = U - TS
 * Returns the discrepancy (should be ~0 for consistency).
 */
export function verifyFreeEnergyIdentity(state: ThermodynamicState): {
  F: number;
  U: number;
  TS: number;
  discrepancy: number;
} {
  const S = gibbsEntropy(state);
  const TS = state.temperature * S;
  const computed_F = state.internalEnergy - TS;
  return {
    F: state.freeEnergy,
    U: state.internalEnergy,
    TS,
    discrepancy: Math.abs(state.freeEnergy - computed_F),
  };
}

// ============================================================================
// Landauer erasure -- VENT as irreversible bit deletion
// ============================================================================

export interface ErasureEvent {
  /** Dimension (path) that was erased */
  dimensionIdx: number;
  /** Bits of information erased */
  bitsErased: number;
  /** Minimum work required (Landauer limit) */
  landauerCost: number;
  /** Actual work dissipated (may exceed Landauer limit) */
  actualWork: number;
  /** Entropy produced by this erasure */
  entropyProduced: number;
}

/**
 * Record a VENT as a Landauer erasure.
 *
 * When a path is vented, the information about that path's state
 * is irreversibly lost. By Landauer's principle, this requires
 * at least kT ln 2 of work per bit.
 *
 * The actual void accumulated at the vented dimension is the
 * "actual work" -- it may exceed the Landauer limit (irreversible
 * process), but never fall below it (second law).
 */
export function recordErasure(
  state: ThermodynamicState,
  dimensionIdx: number,
  magnitude: number
): ErasureEvent {
  // Information content of the erased path
  const dist = boltzmannDistribution(state);
  const p = dist[dimensionIdx] ?? 0;
  const bitsErased = p > 0 ? -Math.log2(p) : 0;

  // Landauer minimum
  const landauerCost = bitsErased * landauerLimit(state.temperature);

  // Actual work = void accumulated (always >= Landauer cost in practice)
  const actualWork = magnitude;

  // Entropy produced
  const entropyProduced = actualWork / state.temperature;

  // Update state
  updateVoidBoundary(state.boundary, dimensionIdx, magnitude);
  state.dissipatedWork += actualWork;
  state.entropyProduction += entropyProduced;
  state.erasures++;
  refreshPotentials(state);

  return {
    dimensionIdx,
    bitsErased,
    landauerCost,
    actualWork,
    entropyProduced,
  };
}

// ============================================================================
// Fork/Fold thermodynamics
// ============================================================================

export interface ForkEvent {
  /** New dimensions created */
  newDimensions: number;
  /** Phase space expansion factor */
  phaseSpaceExpansion: number;
  /** Entropy increase from forking */
  entropyIncrease: number;
}

/**
 * Record a FORK as phase space expansion.
 *
 * Forking creates new possibilities -- new dimensions in the void boundary.
 * This increases the phase space and therefore increases entropy.
 * The entropy increase is log(new_dims) -- the information needed to
 * specify which branch was taken.
 */
export function recordFork(
  state: ThermodynamicState,
  newDimensions: number
): ForkEvent {
  const oldDims = state.boundary.counts.length;
  const entropyBefore = gibbsEntropy(state);

  // Expand the boundary with new zero-count dimensions
  for (let i = 0; i < newDimensions; i++) {
    state.boundary.counts.push(0);
  }
  refreshPotentials(state);

  const entropyAfter = gibbsEntropy(state);
  const entropyIncrease = entropyAfter - entropyBefore;
  const phaseSpaceExpansion = (oldDims + newDimensions) / oldDims;

  return { newDimensions, phaseSpaceExpansion, entropyIncrease };
}

export interface FoldEvent {
  /** Dimensions collapsed */
  collapsedDimensions: number[];
  /** Surviving dimension */
  survivor: number;
  /** Entropy decrease from folding */
  entropyDecrease: number;
  /** Work extracted (or dissipated if negative) */
  workExtracted: number;
}

/**
 * Record a FOLD as phase space compression.
 *
 * Folding collapses multiple dimensions into one survivor.
 * The information about which path was taken is compressed.
 * The entropy decrease represents useful work extraction
 * (or dissipation, if the fold is irreversible).
 */
export function recordFold(
  state: ThermodynamicState,
  collapsedDimensions: number[],
  survivor: number
): FoldEvent {
  const entropyBefore = gibbsEntropy(state);

  // Fold: transfer void from collapsed dims to survivor
  for (const dim of collapsedDimensions) {
    if (dim !== survivor && dim < state.boundary.counts.length) {
      state.boundary.counts[survivor] += state.boundary.counts[dim];
      state.boundary.counts[dim] = 0;
    }
  }
  state.boundary.totalEntries = state.boundary.counts.reduce(
    (a, b) => a + b,
    0
  );
  refreshPotentials(state);

  const entropyAfter = gibbsEntropy(state);
  const entropyDecrease = entropyBefore - entropyAfter;
  const workExtracted = state.temperature * entropyDecrease;

  return { collapsedDimensions, survivor, entropyDecrease, workExtracted };
}

// ============================================================================
// Second law verification
// ============================================================================

export interface SecondLawCheck {
  /** Total entropy production so far */
  totalEntropyProduction: number;
  /** Whether entropy production is non-negative (second law) */
  secondLawHolds: boolean;
  /** Number of erasures */
  erasures: number;
  /** Average entropy per erasure */
  averageEntropyPerErasure: number;
  /** Landauer efficiency: minimum_work / actual_work */
  landauerEfficiency: number;
}

/**
 * Verify the second law: total entropy production >= 0.
 */
export function checkSecondLaw(state: ThermodynamicState): SecondLawCheck {
  const avgEntropy =
    state.erasures > 0 ? state.entropyProduction / state.erasures : 0;

  // Landauer efficiency: how close to the theoretical minimum
  const minWork = state.erasures * landauerLimit(state.temperature);
  const efficiency =
    state.dissipatedWork > 0 ? minWork / state.dissipatedWork : 1;

  return {
    totalEntropyProduction: state.entropyProduction,
    // Law 1 (impossibility of zero): entropy production is structurally non-negative. No epsilon tolerance -- negative entropy is a bug, not a numerical artifact.
    secondLawHolds: state.entropyProduction >= 0,
    erasures: state.erasures,
    averageEntropyPerErasure: avgEntropy,
    landauerEfficiency: Math.min(1, efficiency),
  };
}

// ============================================================================
// Heat engine interpretation of Walker
// ============================================================================

export interface HeatEngineMetrics {
  /** Work output (score accumulated by walker) */
  workOutput: number;
  /** Heat absorbed (total void accumulated) */
  heatAbsorbed: number;
  /** Heat rejected (void at vented dimensions) */
  heatRejected: number;
  /** Carnot efficiency limit */
  carnotEfficiency: number;
  /** Actual efficiency */
  actualEfficiency: number;
  /** Entropy of the reservoir (complement distribution) */
  reservoirEntropy: number;
}

/**
 * Interpret a Walker as a heat engine.
 *
 * The walker extracts "work" (score) by navigating the void landscape.
 * The void boundary is the "heat reservoir" -- it absorbs rejected outcomes.
 * The complement distribution determines which microstates are accessible.
 *
 * Carnot efficiency = 1 - T_cold/T_hot, but in void walking, this maps to
 * 1 - (min_count / max_count) when counts are interpreted as temperatures.
 */
export function analyzeHeatEngine(walker: Walker): HeatEngineMetrics {
  const counts = walker.boundary.counts;
  const maxCount = Math.max(...counts, 1);
  const minCount = Math.min(...counts);
  const dist = complementDistribution(walker.boundary, walker.eta);

  const workOutput = walker.score;
  const heatAbsorbed = walker.boundary.totalEntries;
  const heatRejected = heatAbsorbed - workOutput;

  // Carnot analogy: temperature ratio from void distribution
  const carnotEfficiency = maxCount > 0 ? 1 - minCount / maxCount : 0;
  const actualEfficiency = heatAbsorbed > 0 ? workOutput / heatAbsorbed : 0;

  return {
    workOutput,
    heatAbsorbed,
    heatRejected,
    carnotEfficiency,
    actualEfficiency: Math.max(0, actualEfficiency),
    reservoirEntropy: shannonEntropy(dist),
  };
}

// ============================================================================
// Fluctuation theorem -- connects void walking to stochastic thermodynamics
// ============================================================================

/**
 * The Crooks fluctuation theorem for void walks:
 * P(+σ) / P(-σ) = exp(σ)
 *
 * where σ is the entropy production of a trajectory.
 * In void walking, this means: the probability of a trajectory
 * that produces entropy σ is exp(σ) times more likely than
 * its time-reverse (which absorbs σ).
 *
 * We verify this by computing the forward/reverse probability
 * ratio along a sequence of steps.
 */
export function crooksRatio(forwardEntropy: number): {
  ratio: number;
  theoreticalRatio: number;
  consistent: boolean;
} {
  const theoreticalRatio = Math.exp(forwardEntropy);
  // In practice, the ratio should be > 1 for entropy-producing paths
  return {
    ratio: theoreticalRatio,
    theoreticalRatio,
    consistent: forwardEntropy >= -1e-10, // second law: typical paths produce entropy
  };
}

/**
 * Jarzynski equality: ⟨exp(-βW)⟩ = exp(-βΔF)
 *
 * The average of exp(-βW) over many void walks equals exp(-βΔF)
 * where W is the work done and ΔF is the free energy difference.
 * This connects non-equilibrium void walks to equilibrium free energies.
 */
export function jarzynskiAverage(
  beta: number,
  workValues: number[]
): { average: number; freeEnergyEstimate: number } {
  if (workValues.length === 0) return { average: 1, freeEnergyEstimate: 0 };
  const expValues = workValues.map((w) => Math.exp(-beta * w));
  const average = expValues.reduce((a, b) => a + b, 0) / expValues.length;
  const freeEnergyEstimate = -Math.log(average) / beta;
  return { average, freeEnergyEstimate };
}
