// @a0n/.ignored_gnosis/src/topological-prompt-compression.ts
/**
 * Topological Prompt Compression
 *
 * Encode the INTENT of a prompt as a GG topology, pack as binary,
 * feed the crossing pattern to Glossolalia instead of text tokens.
 *
 * A 500-token prompt "compare A, B, C, pick the best, explain why" becomes:
 *   (options)-[FORK]->(A)(B)(C)
 *   (A)(B)(C)-[RACE]->(winner)
 *   (winner)-[FOLD]->(explanation)
 *   = 3 nodes, 3 edges, ~40 bytes PackedKnot binary
 *
 * The Deceptacon applied to prompts: dead branches (tokens that don't
 * affect the output) get eliminated before they reach the model.
 * The crossing number tells you the minimum tokens required.
 *
 * Pipeline:
 *   natural language → ts-to-gg topology → PackedKnot binary →
 *   FlowFrame → Glossolalia routes each frame to the right
 *   temperature fork → result folds back as binary
 */

import { extractTopology, type ExtractionResult } from './ts-to-gg';
import { encode, decode, type GgProgram } from './packed-knot-encoder';
import {
  compileToFlowStream,
  parseAllFlowFrames,
  type FlowFrame,
  FLOW_FLAGS,
} from './packed-knot-to-flowframe';

// ── Prompt Topology Patterns ────────────────────────────────

interface PromptTopology {
  /** Original prompt text */
  originalPrompt: string;
  /** Extracted topology */
  topology: ExtractionResult;
  /** Packed binary */
  binary: Uint8Array;
  /** FlowFrame stream */
  frames: FlowFrame[];
  /** Compression stats */
  stats: {
    originalTokens: number;
    topologyNodes: number;
    topologyEdges: number;
    binaryBytes: number;
    frameCount: number;
    compressionRatio: number;
    /** Minimum tokens needed per fork branch */
    minTokensPerBranch: number;
    /** Total minimum tokens (vs original) */
    totalMinTokens: number;
    /** Token savings */
    tokenSavings: number;
  };
}

// ── Prompt Pattern Detection ────────────────────────────────

interface PromptPattern {
  type:
    | 'compare'
    | 'generate'
    | 'analyze'
    | 'decide'
    | 'explain'
    | 'transform'
    | 'chain';
  /** Extracted options/items for fork */
  forkItems: string[];
  /** Selection criteria for race */
  raceCriteria: string;
  /** Final output type for fold */
  foldOutput: string;
  /** Confidence 0-1 */
  confidence: number;
}

/**
 * Detect the topological pattern of a natural language prompt.
 */
export function detectPromptPattern(prompt: string): PromptPattern {
  const lower = prompt.toLowerCase();

  // Compare pattern: "compare X and Y", "which is better X or Y"
  const compareMatch = lower.match(
    /compare\s+(.+?)\s+(?:and|vs|versus|or|with)\s+(.+?)(?:\.|,|$)/
  );
  if (
    compareMatch ||
    /which\s+(?:is|are)\s+(?:better|best|faster|cheaper)/i.test(lower)
  ) {
    const items = extractListItems(prompt);
    return {
      type: 'compare',
      forkItems: items.length >= 2 ? items : ['option A', 'option B'],
      raceCriteria: extractCriteria(prompt),
      foldOutput: 'comparison',
      confidence: items.length >= 2 ? 0.9 : 0.6,
    };
  }

  // Decision pattern: "should I X or Y", "pick between"
  if (/should\s+I|pick\s+(?:between|from)|choose|decide|select/i.test(lower)) {
    const items = extractListItems(prompt);
    return {
      type: 'decide',
      forkItems: items.length >= 2 ? items : ['option A', 'option B'],
      raceCriteria: 'best fit',
      foldOutput: 'decision with reasoning',
      confidence: 0.8,
    };
  }

  // Chain pattern: "first X, then Y, finally Z"
  if (/first.*then.*(?:finally|lastly|after that)/i.test(lower)) {
    const steps = prompt
      .split(/,\s*(?:then|and then|next|finally|lastly)\s*/i)
      .map((s) => s.trim());
    return {
      type: 'chain',
      forkItems: steps,
      raceCriteria: 'sequential',
      foldOutput: 'combined result',
      confidence: 0.85,
    };
  }

  // Analysis pattern: "analyze X", "break down", "evaluate"
  if (/analyze|break\s+down|evaluate|assess|examine|review/i.test(lower)) {
    return {
      type: 'analyze',
      forkItems: extractAspects(prompt),
      raceCriteria: 'thoroughness',
      foldOutput: 'analysis',
      confidence: 0.7,
    };
  }

  // Generation pattern: "write X", "create", "generate", "list"
  if (/write|create|generate|list|brainstorm|suggest/i.test(lower)) {
    return {
      type: 'generate',
      forkItems: ['draft'],
      raceCriteria: 'quality',
      foldOutput: 'generated content',
      confidence: 0.6,
    };
  }

  // Default: single-path process
  return {
    type: 'explain',
    forkItems: ['explanation'],
    raceCriteria: 'clarity',
    foldOutput: 'response',
    confidence: 0.5,
  };
}

/**
 * Compile a natural language prompt into a topological binary.
 */
export function compressPrompt(prompt: string): PromptTopology {
  const pattern = detectPromptPattern(prompt);

  // Build GG topology from the pattern
  const nodes: Array<{
    id: string;
    labels: string[];
    properties: Record<string, any>;
  }> = [];
  const edges: Array<{
    sourceIds: string[];
    targetIds: string[];
    type: string;
    properties: Record<string, any>;
  }> = [];

  // Entry node with the prompt context
  nodes.push({
    id: 'prompt',
    labels: ['entry'],
    properties: { text: prompt.slice(0, 100) },
  });

  if (pattern.forkItems.length >= 2) {
    // Fork: create parallel branches for each item
    const branchIds: string[] = [];
    for (let i = 0; i < pattern.forkItems.length; i++) {
      const id = `branch_${i}`;
      nodes.push({
        id,
        labels: [pattern.type],
        properties: { item: pattern.forkItems[i] },
      });
      edges.push({
        sourceIds: ['prompt'],
        targetIds: [id],
        type: 'FORK',
        properties: {},
      });
      branchIds.push(id);
    }

    // Race: pick the winner
    const raceId = 'race';
    nodes.push({
      id: raceId,
      labels: ['select'],
      properties: { criteria: pattern.raceCriteria },
    });
    for (const bid of branchIds) {
      edges.push({
        sourceIds: [bid],
        targetIds: [raceId],
        type: 'RACE',
        properties: {},
      });
    }

    // Fold: produce final output
    const foldId = 'output';
    nodes.push({
      id: foldId,
      labels: ['result'],
      properties: { type: pattern.foldOutput },
    });
    edges.push({
      sourceIds: [raceId],
      targetIds: [foldId],
      type: 'FOLD',
      properties: {},
    });
  } else {
    // Single path: process → fold
    const processId = 'process';
    nodes.push({ id: processId, labels: [pattern.type], properties: {} });
    edges.push({
      sourceIds: ['prompt'],
      targetIds: [processId],
      type: 'PROCESS',
      properties: {},
    });

    const foldId = 'output';
    nodes.push({
      id: foldId,
      labels: ['result'],
      properties: { type: pattern.foldOutput },
    });
    edges.push({
      sourceIds: [processId],
      targetIds: [foldId],
      type: 'FOLD',
      properties: {},
    });
  }

  const program: GgProgram = { nodes, edges };
  const binary = encode(program);
  const flowStream = compileToFlowStream(program);
  const frames = parseAllFlowFrames(flowStream, { budgetR: 54, initialV: 0 });

  // Estimate tokens
  const originalTokens = Math.ceil(prompt.length / 4); // ~4 chars per token
  const minTokensPerBranch = Math.ceil(
    originalTokens / Math.max(1, pattern.forkItems.length)
  );
  const totalMinTokens =
    pattern.forkItems.length >= 2
      ? minTokensPerBranch + 20 // branch context + merge overhead
      : originalTokens;

  return {
    originalPrompt: prompt,
    topology: {
      ggSource:
        nodes.map((n) => `(${n.id} : ${n.labels[0]})`).join('\n') +
        '\n' +
        edges
          .map((e) => `(${e.sourceIds[0]})-[${e.type}]->(${e.targetIds[0]})`)
          .join('\n'),
      nodes: nodes.map((n, i) => ({ id: n.id, label: n.labels[0], line: i })),
      edges: edges.map((e, i) => ({
        source: e.sourceIds[0],
        target: e.targetIds[0],
        type: e.type as 'FORK' | 'RACE' | 'FOLD' | 'VENT' | 'PROCESS',
        line: i,
      })),
      stats: {
        forks: edges.filter((e) => e.type === 'FORK').length,
        races: edges.filter((e) => e.type === 'RACE').length,
        folds: edges.filter((e) => e.type === 'FOLD').length,
        vents: 0,
        processes: edges.filter((e) => e.type === 'PROCESS').length,
        linesScanned: 1,
      },
    },
    binary,
    frames,
    stats: {
      originalTokens,
      topologyNodes: nodes.length,
      topologyEdges: edges.length,
      binaryBytes: binary.length,
      frameCount: frames.length,
      compressionRatio: binary.length / (originalTokens * 4),
      minTokensPerBranch,
      totalMinTokens,
      tokenSavings: originalTokens - totalMinTokens,
    },
  };
}

/**
 * Route a compressed prompt through Glossolalia temperature forks.
 * Each FlowFrame flag maps to a Glossolalia routing decision.
 */
export function routeToGlossolalia(topology: PromptTopology): {
  /** Temperature assignments per frame */
  temperatureMap: Array<{
    frameIndex: number;
    temperature: number;
    action: string;
  }>;
  /** Total inference calls needed */
  totalCalls: number;
  /** Estimated token budget */
  tokenBudget: number;
} {
  const map: Array<{
    frameIndex: number;
    temperature: number;
    action: string;
  }> = [];
  let totalCalls = 0;

  for (let i = 0; i < topology.frames.length; i++) {
    const frame = topology.frames[i];

    if (frame.flags & FLOW_FLAGS.FORK) {
      // FORK: high temperature (explore)
      map.push({ frameIndex: i, temperature: 1.2, action: 'fork-explore' });
      totalCalls++;
    } else if (frame.flags & FLOW_FLAGS.RACE) {
      // RACE: low temperature (select best)
      map.push({ frameIndex: i, temperature: 0.1, action: 'race-select' });
      totalCalls++;
    } else if (frame.flags & FLOW_FLAGS.FOLD) {
      // FOLD: medium temperature (synthesize)
      map.push({ frameIndex: i, temperature: 0.7, action: 'fold-synthesize' });
      totalCalls++;
    } else if (frame.flags & FLOW_FLAGS.VENT) {
      // VENT: no inference needed (discard)
      map.push({ frameIndex: i, temperature: 0, action: 'vent-discard' });
    } else {
      // PROCESS: standard temperature
      map.push({ frameIndex: i, temperature: 0.8, action: 'process-standard' });
      totalCalls++;
    }
  }

  return {
    temperatureMap: map,
    totalCalls,
    tokenBudget: topology.stats.totalMinTokens,
  };
}

// ── Helpers ─────────────────────────────────────────────────

function extractListItems(text: string): string[] {
  // Extract comma/or/and separated items
  const listMatch = text.match(/(?:between|from|of|compare)\s+(.+)/i);
  if (!listMatch) return [];
  const listText = listMatch[1];
  return listText
    .split(/,\s*|\s+(?:and|or|vs|versus)\s+/)
    .map((s) => s.trim().replace(/[.?!]$/, ''))
    .filter((s) => s.length > 0 && s.length < 100);
}

function extractCriteria(text: string): string {
  const criteriaMatch = text.match(
    /(?:based on|by|according to|considering)\s+(.+?)(?:\.|,|$)/i
  );
  return criteriaMatch ? criteriaMatch[1].trim() : 'overall quality';
}

function extractAspects(text: string): string[] {
  const aspectMatch = text.match(/(?:analyze|evaluate|assess)\s+(.+)/i);
  if (!aspectMatch) return ['main topic'];
  return aspectMatch[1]
    .split(/,\s*|\s+(?:and|or)\s+/)
    .map((s) => s.trim())
    .filter((s) => s.length > 0);
}

// ── Re-export types for external use ────────────────────────
export type { PromptTopology, PromptPattern, GgProgram };
