// @a0n/.ignored_gnosis/src/transformation-family-formalization.ts
import { decompose, type TransformObject } from './decomposition-topology';
import {
  reinforce,
  type ReinforcementAttempt,
  type ReinforcementOutcome,
} from './reinforcement-topology';

export const TRANSFORMATION_FAMILY_THEOREM_IDS = [
  'THM-TRANSFORMATION-PROJECTION-ADDS-STRUCTURE',
  'THM-TRANSFORMATION-REINFORCEMENT-PRESERVES-STRUCTURE',
  'THM-TRANSFORMATION-DECOMPOSITION-NONINCREASING',
  'THM-TRANSFORMATION-PROJECTION-REINFORCEMENT-EXCLUSIVE',
  'THM-TRANSFORMATION-REINFORCEMENT-DECOMPOSITION-EXCLUSIVE',
  'THM-TRANSFORMATION-FEATURELESS-CANNOT-REINFORCE',
  'THM-TRANSFORMATION-DECOMPOSE-THEN-REINFORCE-REQUIRES-RESTORATION',
] as const;

export type TransformationFamilyTheoremId =
  (typeof TRANSFORMATION_FAMILY_THEOREM_IDS)[number];

export type TransformationFamily =
  | 'projection'
  | 'reinforcement'
  | 'decomposition';

export type TransformationFamilyClassification =
  | TransformationFamily
  | 'unclassified';

export interface TransformationFamilySignal {
  readonly id: string;
  readonly beforeCrossings: number;
  readonly afterCrossings: number;
  readonly beforeWeight: number;
  readonly afterWeight: number;
}

export interface TransformationFamilyFormalizationInput {
  readonly signals: readonly TransformationFamilySignal[];
  readonly generatedAt?: string;
  readonly source?: string;
}

export interface TransformationFamilyEntry {
  readonly id: string;
  readonly family: TransformationFamilyClassification;
  readonly beforeCrossings: number;
  readonly afterCrossings: number;
  readonly beforeWeight: number;
  readonly afterWeight: number;
  readonly crossingsDelta: number;
  readonly weightDelta: number;
  readonly featurelessBefore: boolean;
  readonly featurelessAfter: boolean;
  readonly validProjection: boolean;
  readonly validReinforcement: boolean;
  readonly validDecomposition: boolean;
  readonly restorationRequired: boolean;
}

export interface TransformationFamilyAggregate {
  readonly total: number;
  readonly projectionCount: number;
  readonly reinforcementCount: number;
  readonly decompositionCount: number;
  readonly unclassifiedCount: number;
  readonly featurelessBeforeCount: number;
  readonly featurelessAfterCount: number;
  readonly restorationRequiredCount: number;
}

export interface TransformationFamilyBundle {
  readonly generatedAt: string;
  readonly source: string;
  readonly theoremIds: readonly TransformationFamilyTheoremId[];
  readonly entries: readonly TransformationFamilyEntry[];
  readonly aggregate: TransformationFamilyAggregate;
}

export interface ReinforcementTransformationAttemptInput {
  readonly id: string;
  readonly attempt: ReinforcementAttempt;
}

export type ReinforcementTransformationBlockedReason = 'fizzle' | 'shatter';

export interface ReinforcementTransformationAttemptEntry {
  readonly id: string;
  readonly outcome: ReinforcementOutcome['kind'];
  readonly beforeCrossings: number;
  readonly beforeWeight: number;
  readonly emittedSignal: boolean;
  readonly restorationRequired: boolean;
  readonly blockedReason?: ReinforcementTransformationBlockedReason;
  readonly signal?: TransformationFamilySignal;
}

export interface ReinforcementTransformationAttemptAggregate {
  readonly totalAttempts: number;
  readonly successCount: number;
  readonly fizzleCount: number;
  readonly shatterCount: number;
  readonly emittedSignalCount: number;
  readonly restorationRequiredCount: number;
}

export interface ReinforcementTransformationBridgeBundle {
  readonly generatedAt: string;
  readonly source: string;
  readonly attempts: readonly ReinforcementTransformationAttemptEntry[];
  readonly aggregate: ReinforcementTransformationAttemptAggregate;
  readonly formalization: TransformationFamilyBundle;
}

export interface DecompositionTransformationAttemptInput {
  readonly id: string;
  readonly object: TransformObject;
  readonly budget: number;
}

export interface DecompositionTransformationAttemptEntry {
  readonly id: string;
  readonly budget: number;
  readonly beforeCrossings: number;
  readonly afterCrossings: number;
  readonly beforeWeight: number;
  readonly afterWeight: number;
  readonly featurelessBefore: boolean;
  readonly featurelessAfter: boolean;
  readonly fullyDecomposed: boolean;
  readonly effectiveDecomposition: boolean;
  readonly signal: TransformationFamilySignal;
}

export interface DecompositionTransformationAttemptAggregate {
  readonly totalAttempts: number;
  readonly effectiveDecompositionCount: number;
  readonly noOpCount: number;
  readonly fullyDecomposedCount: number;
  readonly featurelessBeforeCount: number;
  readonly featurelessAfterCount: number;
}

export interface DecompositionTransformationBridgeBundle {
  readonly generatedAt: string;
  readonly source: string;
  readonly attempts: readonly DecompositionTransformationAttemptEntry[];
  readonly aggregate: DecompositionTransformationAttemptAggregate;
  readonly formalization: TransformationFamilyBundle;
}

interface NormalizedSignal {
  readonly id: string;
  readonly beforeCrossings: number;
  readonly afterCrossings: number;
  readonly beforeWeight: number;
  readonly afterWeight: number;
}

function nowIso(): string {
  return new Date().toISOString();
}

function normalizeNonNegativeInt(value: number): number {
  if (!Number.isFinite(value)) {
    return 0;
  }
  return Math.max(0, Math.floor(value));
}

function normalizeId(value: string, index: number): string {
  const trimmed = value.trim();
  return trimmed.length > 0 ? trimmed : `signal-${index}`;
}

function normalizeSignal(
  signal: TransformationFamilySignal,
  index: number
): NormalizedSignal {
  return {
    id: normalizeId(signal.id, index),
    beforeCrossings: normalizeNonNegativeInt(signal.beforeCrossings),
    afterCrossings: normalizeNonNegativeInt(signal.afterCrossings),
    beforeWeight: normalizeNonNegativeInt(signal.beforeWeight),
    afterWeight: normalizeNonNegativeInt(signal.afterWeight),
  };
}

function compareSignals(
  left: NormalizedSignal,
  right: NormalizedSignal
): number {
  return (
    left.id.localeCompare(right.id) ||
    left.beforeCrossings - right.beforeCrossings ||
    left.afterCrossings - right.afterCrossings ||
    left.beforeWeight - right.beforeWeight ||
    left.afterWeight - right.afterWeight
  );
}

function classifyFamily(
  signal: NormalizedSignal
): TransformationFamilyClassification {
  if (signal.beforeCrossings === 0 && signal.afterCrossings > 0) {
    return 'projection';
  }
  if (signal.afterCrossings < signal.beforeCrossings) {
    return 'decomposition';
  }
  if (
    signal.beforeCrossings > 0 &&
    signal.beforeCrossings === signal.afterCrossings &&
    signal.afterWeight > signal.beforeWeight
  ) {
    return 'reinforcement';
  }
  return 'unclassified';
}

function toEntry(signal: NormalizedSignal): TransformationFamilyEntry {
  const family = classifyFamily(signal);
  const crossingsDelta = signal.afterCrossings - signal.beforeCrossings;
  const weightDelta = signal.afterWeight - signal.beforeWeight;
  const featurelessBefore = signal.beforeCrossings === 0;
  const featurelessAfter = signal.afterCrossings === 0;
  const validProjection = family === 'projection';
  const validReinforcement = family === 'reinforcement';
  const validDecomposition = family === 'decomposition';
  const restorationRequired =
    featurelessBefore &&
    signal.afterCrossings === signal.beforeCrossings &&
    signal.afterWeight > signal.beforeWeight;

  return {
    id: signal.id,
    family,
    beforeCrossings: signal.beforeCrossings,
    afterCrossings: signal.afterCrossings,
    beforeWeight: signal.beforeWeight,
    afterWeight: signal.afterWeight,
    crossingsDelta,
    weightDelta,
    featurelessBefore,
    featurelessAfter,
    validProjection,
    validReinforcement,
    validDecomposition,
    restorationRequired,
  };
}

/**
 * Builds the Transformation Family Signal From Reinforcement Attempt.
 */
export function buildTransformationFamilySignalFromReinforcementAttempt(
  input: ReinforcementTransformationAttemptInput
): TransformationFamilySignal | undefined {
  const id = normalizeId(input.id, 0);
  const outcome = reinforce(input.attempt);
  if (outcome.kind !== 'success') {
    return undefined;
  }

  return {
    id,
    beforeCrossings: input.attempt.target.crossings,
    afterCrossings: input.attempt.target.crossings,
    beforeWeight: input.attempt.target.weight,
    afterWeight: outcome.newWeight,
  };
}

/**
 * Builds the Transformation Family Signal From Decomposition Attempt.
 */
export function buildTransformationFamilySignalFromDecompositionAttempt(
  input: DecompositionTransformationAttemptInput
): TransformationFamilySignal {
  const id = normalizeId(input.id, 0);
  const result = decompose(input.object, input.budget);

  return {
    id,
    beforeCrossings: input.object.crossings,
    afterCrossings: result.residualCrossings,
    beforeWeight: input.object.weight,
    afterWeight: result.residualWeight,
  };
}

function toReinforcementAttemptEntry(
  input: ReinforcementTransformationAttemptInput,
  index: number
): ReinforcementTransformationAttemptEntry {
  const id = normalizeId(input.id, index);
  const outcome = reinforce(input.attempt);
  const signal = buildTransformationFamilySignalFromReinforcementAttempt({
    id,
    attempt: input.attempt,
  });
  const restorationRequired =
    outcome.kind === 'fizzle' &&
    input.attempt.energy > 0 &&
    input.attempt.target.crossings === 0;

  return {
    id,
    outcome: outcome.kind,
    beforeCrossings: input.attempt.target.crossings,
    beforeWeight: input.attempt.target.weight,
    emittedSignal: signal !== undefined,
    restorationRequired,
    ...(outcome.kind !== 'success' ? { blockedReason: outcome.kind } : {}),
    ...(signal !== undefined ? { signal } : {}),
  };
}

function toDecompositionAttemptEntry(
  input: DecompositionTransformationAttemptInput,
  index: number
): DecompositionTransformationAttemptEntry {
  const id = normalizeId(input.id, index);
  const signal = buildTransformationFamilySignalFromDecompositionAttempt({
    id,
    object: input.object,
    budget: input.budget,
  });

  return {
    id,
    budget: normalizeNonNegativeInt(input.budget),
    beforeCrossings: signal.beforeCrossings,
    afterCrossings: signal.afterCrossings,
    beforeWeight: signal.beforeWeight,
    afterWeight: signal.afterWeight,
    featurelessBefore: signal.beforeCrossings === 0,
    featurelessAfter: signal.afterCrossings === 0,
    fullyDecomposed: signal.beforeCrossings > 0 && signal.afterCrossings === 0,
    effectiveDecomposition: signal.afterCrossings < signal.beforeCrossings,
    signal,
  };
}

/**
 * Handles the gnosis derive Transformation Family Formalization workflow.
 */
export function deriveTransformationFamilyFormalization(
  input: TransformationFamilyFormalizationInput
): TransformationFamilyBundle {
  const normalized = input.signals
    .map((signal, index) => normalizeSignal(signal, index))
    .sort(compareSignals);
  const entries = normalized.map(toEntry);

  const aggregate: TransformationFamilyAggregate = {
    total: entries.length,
    projectionCount: entries.filter((entry) => entry.family === 'projection')
      .length,
    reinforcementCount: entries.filter(
      (entry) => entry.family === 'reinforcement'
    ).length,
    decompositionCount: entries.filter(
      (entry) => entry.family === 'decomposition'
    ).length,
    unclassifiedCount: entries.filter(
      (entry) => entry.family === 'unclassified'
    ).length,
    featurelessBeforeCount: entries.filter((entry) => entry.featurelessBefore)
      .length,
    featurelessAfterCount: entries.filter((entry) => entry.featurelessAfter)
      .length,
    restorationRequiredCount: entries.filter(
      (entry) => entry.restorationRequired
    ).length,
  };

  return {
    generatedAt: input.generatedAt ?? nowIso(),
    source: input.source ?? 'transformationFamilyFormalization/v1',
    theoremIds: TRANSFORMATION_FAMILY_THEOREM_IDS,
    entries,
    aggregate,
  };
}

/**
 * Handles the gnosis derive Transformation Family From Reinforcement Attempts workflow.
 */
export function deriveTransformationFamilyFromReinforcementAttempts(input: {
  readonly attempts: readonly ReinforcementTransformationAttemptInput[];
  readonly generatedAt?: string;
  readonly source?: string;
}): ReinforcementTransformationBridgeBundle {
  const attempts = input.attempts.map(toReinforcementAttemptEntry);
  const signals = attempts.flatMap((entry) =>
    entry.signal === undefined ? [] : [entry.signal]
  );
  const source = input.source ?? 'transformationFamilyFromReinforcement/v1';
  const formalization = deriveTransformationFamilyFormalization({
    signals,
    ...(input.generatedAt !== undefined ? { generatedAt: input.generatedAt } : {}),
    source,
  });

  const aggregate: ReinforcementTransformationAttemptAggregate = {
    totalAttempts: attempts.length,
    successCount: attempts.filter((entry) => entry.outcome === 'success')
      .length,
    fizzleCount: attempts.filter((entry) => entry.outcome === 'fizzle').length,
    shatterCount: attempts.filter((entry) => entry.outcome === 'shatter')
      .length,
    emittedSignalCount: attempts.filter((entry) => entry.emittedSignal).length,
    restorationRequiredCount: attempts.filter(
      (entry) => entry.restorationRequired
    ).length,
  };

  return {
    generatedAt: formalization.generatedAt,
    source,
    attempts,
    aggregate,
    formalization,
  };
}

/**
 * Handles the gnosis derive Transformation Family From Decomposition Attempts workflow.
 */
export function deriveTransformationFamilyFromDecompositionAttempts(input: {
  readonly attempts: readonly DecompositionTransformationAttemptInput[];
  readonly generatedAt?: string;
  readonly source?: string;
}): DecompositionTransformationBridgeBundle {
  const attempts = input.attempts.map(toDecompositionAttemptEntry);
  const signals = attempts.map((entry) => entry.signal);
  const source = input.source ?? 'transformationFamilyFromDecomposition/v1';
  const formalization = deriveTransformationFamilyFormalization({
    signals,
    ...(input.generatedAt !== undefined ? { generatedAt: input.generatedAt } : {}),
    source,
  });

  const aggregate: DecompositionTransformationAttemptAggregate = {
    totalAttempts: attempts.length,
    effectiveDecompositionCount: attempts.filter(
      (entry) => entry.effectiveDecomposition
    ).length,
    noOpCount: attempts.filter((entry) => !entry.effectiveDecomposition).length,
    fullyDecomposedCount: attempts.filter((entry) => entry.fullyDecomposed)
      .length,
    featurelessBeforeCount: attempts.filter((entry) => entry.featurelessBefore)
      .length,
    featurelessAfterCount: attempts.filter((entry) => entry.featurelessAfter)
      .length,
  };

  return {
    generatedAt: formalization.generatedAt,
    source,
    attempts,
    aggregate,
    formalization,
  };
}
