// @a0n/.ignored_gnosis/src/triangulation-dimensional-coercion.ts
import { type ReductionProjectionInput } from './spiderweb-topology';

export interface TriMesh {
  readonly vertices: number;
  readonly edges: number;
  readonly faces: number;
}

export interface Realization {
  readonly ambientDim: number;
  readonly mesh: TriMesh;
}

export interface SphericalTriangulation {
  readonly mesh: TriMesh;
}

export interface TriangulationCoercionEdge {
  readonly id: string;
  readonly sourceDim: number;
  readonly targetDim: number;
  readonly realization: Realization;
  readonly roundTrip: Realization;
  readonly unknotDebtBefore: number;
  readonly unknotDebtAfter: number;
  readonly unknotDebtDrop: number;
  readonly strictlyReducesUnknotDebt: boolean;
  readonly eulerTwoWitnessed: boolean;
  readonly preservesMesh: boolean;
  readonly preservesSimplexMass: boolean;
  readonly roundTripPreservesMesh: boolean;
}

export interface TriangulationCoercionMatrix {
  readonly dimensions: readonly number[];
  readonly edges: readonly TriangulationCoercionEdge[];
}

function toNat(value: number): number {
  if (!Number.isFinite(value)) return 0;
  return Math.max(0, Math.trunc(value));
}

function normalizeDim(value: number): number {
  return Math.max(1, toNat(value));
}

function meshEquals(a: TriMesh, b: TriMesh): boolean {
  return a.vertices === b.vertices && a.edges === b.edges && a.faces === b.faces;
}

function uniqueDimensions(dimensions: readonly number[]): number[] {
  const out: number[] = [];
  const seen = new Set<number>();
  for (const dim of dimensions) {
    const normalized = normalizeDim(dim);
    if (seen.has(normalized)) continue;
    seen.add(normalized);
    out.push(normalized);
  }
  return out;
}

/**
 * Handles the gnosis simplex Mass workflow.
 */
export function simplexMass(mesh: TriMesh): number {
  return toNat(mesh.vertices) + toNat(mesh.edges) + toNat(mesh.faces);
}

/**
 * Handles the gnosis euler Two workflow.
 */
export function eulerTwo(mesh: TriMesh): boolean {
  return toNat(mesh.vertices) + toNat(mesh.faces) === toNat(mesh.edges) + 2;
}

/**
 * Handles the gnosis in Dimension workflow.
 */
export function inDimension(mesh: TriMesh, ambientDim: number): Realization {
  return { ambientDim: normalizeDim(ambientDim), mesh };
}

/**
 * Handles the gnosis lift To Max Dim workflow.
 */
export function liftToMaxDim(
  sourceDim: number,
  targetDim: number,
  realization: Realization
): Realization {
  return {
    ambientDim: Math.max(normalizeDim(sourceDim), normalizeDim(targetDim)),
    mesh: realization.mesh,
  };
}

/**
 * Handles the gnosis project Dim workflow.
 */
export function projectDim(targetDim: number, realization: Realization): Realization {
  return {
    ambientDim: normalizeDim(targetDim),
    mesh: realization.mesh,
  };
}

/**
 * Handles the gnosis coerce Across Dimensions workflow.
 */
export function coerceAcrossDimensions(
  mesh: TriMesh,
  sourceDim: number,
  targetDim: number
): Realization {
  return projectDim(
    targetDim,
    liftToMaxDim(sourceDim, targetDim, inDimension(mesh, sourceDim))
  );
}

/**
 * Handles the gnosis round Trip Across Dimensions workflow.
 */
export function roundTripAcrossDimensions(
  mesh: TriMesh,
  sourceDim: number,
  targetDim: number
): Realization {
  const forward = coerceAcrossDimensions(mesh, sourceDim, targetDim);
  return coerceAcrossDimensions(forward.mesh, targetDim, sourceDim);
}

/**
 * Handles the gnosis unknot Debt To Target workflow.
 */
export function unknotDebtToTarget(sourceDim: number, targetDim: number): number {
  const source = normalizeDim(sourceDim);
  const target = normalizeDim(targetDim);
  return Math.max(0, target - source);
}

/**
 * Handles the gnosis unknot Debt After Coercion workflow.
 */
export function unknotDebtAfterCoercion(
  sourceDim: number,
  targetDim: number
): number {
  void normalizeDim(sourceDim);
  void normalizeDim(targetDim);
  /* Placeholder: coercion realizes the mesh in the target ambient space; debt
     drop is modeled by unknotDebtDrop / matrix edges, not a simple dim delta. */
  return 0;
}

/**
 * Handles the gnosis unknot Debt Drop workflow.
 */
export function unknotDebtDrop(sourceDim: number, targetDim: number): number {
  const before = unknotDebtToTarget(sourceDim, targetDim);
  const after = unknotDebtAfterCoercion(sourceDim, targetDim);
  return Math.max(0, before - after);
}

/**
 * Handles the gnosis sphere To Triangulation workflow.
 */
export function sphereToTriangulation(surface: SphericalTriangulation): TriMesh {
  return surface.mesh;
}

/**
 * Handles the gnosis triangulation To Sphere workflow.
 */
export function triangulationToSphere(mesh: TriMesh): SphericalTriangulation {
  if (!eulerTwo(mesh)) {
    throw new RangeError('triangulationToSphere requires an Euler-2 triangular mesh witness.');
  }
  return { mesh };
}

/**
 * Handles the gnosis tri Sphere Tri Round Trip workflow.
 */
export function triSphereTriRoundTrip(mesh: TriMesh): TriMesh {
  return sphereToTriangulation(triangulationToSphere(mesh));
}

/**
 * Handles the gnosis sphere Tri Sphere Round Trip workflow.
 */
export function sphereTriSphereRoundTrip(
  surface: SphericalTriangulation
): SphericalTriangulation {
  return triangulationToSphere(sphereToTriangulation(surface));
}

function buildEdgeId(idPrefix: string, sourceDim: number, targetDim: number): string {
  return `${idPrefix}:${sourceDim}->${targetDim}`;
}

/**
 * Builds the Triangulation Coercion Edge.
 */
export function buildTriangulationCoercionEdge(
  mesh: TriMesh,
  sourceDim: number,
  targetDim: number,
  idPrefix: string = 'triangulation'
): TriangulationCoercionEdge {
  const normalizedSourceDim = normalizeDim(sourceDim);
  const normalizedTargetDim = normalizeDim(targetDim);
  const realization = coerceAcrossDimensions(mesh, sourceDim, targetDim);
  const roundTrip = roundTripAcrossDimensions(mesh, sourceDim, targetDim);
  const debtBefore = unknotDebtToTarget(normalizedSourceDim, normalizedTargetDim);
  const debtAfter = unknotDebtAfterCoercion(
    normalizedSourceDim,
    normalizedTargetDim
  );
  const debtDrop = Math.max(0, debtBefore - debtAfter);
  return {
    id: buildEdgeId(idPrefix, normalizedSourceDim, normalizedTargetDim),
    sourceDim: normalizedSourceDim,
    targetDim: normalizedTargetDim,
    realization,
    roundTrip,
    unknotDebtBefore: debtBefore,
    unknotDebtAfter: debtAfter,
    unknotDebtDrop: debtDrop,
    strictlyReducesUnknotDebt: debtDrop > 0,
    eulerTwoWitnessed: eulerTwo(mesh),
    preservesMesh: meshEquals(realization.mesh, mesh),
    preservesSimplexMass: simplexMass(realization.mesh) === simplexMass(mesh),
    roundTripPreservesMesh: meshEquals(roundTrip.mesh, mesh),
  };
}

/**
 * Builds the Triangulation Coercion Matrix.
 */
export function buildTriangulationCoercionMatrix(
  mesh: TriMesh,
  dimensions: readonly number[],
  idPrefix: string = 'triangulation'
): TriangulationCoercionMatrix {
  const normalized = uniqueDimensions(dimensions);
  const edges: TriangulationCoercionEdge[] = [];
  for (const sourceDim of normalized) {
    for (const targetDim of normalized) {
      if (sourceDim === targetDim) continue;
      edges.push(
        buildTriangulationCoercionEdge(mesh, sourceDim, targetDim, idPrefix)
      );
    }
  }
  return {
    dimensions: normalized,
    edges,
  };
}

/**
 * Handles the gnosis coercion Edge To Reduction Input workflow.
 */
export function coercionEdgeToReductionInput(
  edge: TriangulationCoercionEdge
): ReductionProjectionInput {
  const observedDim = edge.sourceDim;
  const latentDim = edge.targetDim;
  const hasGap = edge.strictlyReducesUnknotDebt;
  const hasLatentSolution = edge.eulerTwoWitnessed && edge.roundTripPreservesMesh;

  return {
    id: edge.id,
    frame: {
      observedDim,
      latentDim,
    },
    witness: {
      observedHasSolution: !hasGap,
      latentHasSolution: hasLatentSolution,
      hasCollapsedSplit: hasGap && hasLatentSolution,
    },
  };
}

/**
 * Handles the gnosis coercion Matrix To Reduction Inputs workflow.
 */
export function coercionMatrixToReductionInputs(
  matrix: TriangulationCoercionMatrix
): ReductionProjectionInput[] {
  return matrix.edges.map((edge) => coercionEdgeToReductionInput(edge));
}
