// @a0n/.ignored_gnosis/src/ts-bridge.test.ts
import { spawnSync } from 'node:child_process';
import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import { pathToFileURL } from 'node:url';
import { checkGgProgram } from '@a0n/aeon-logic';
import { describe, expect, it } from '@a0n/gnosis/test';
import { QDoc } from './crdt/index';
import {
  createBraidFoldCertificate,
  selectCertifiedBraidFastPath,
} from './mesh-policy';
import { GnosisCoreCache } from './runtime/core-cache';
import {
  compileTypeScriptToGnosis,
  deserializeTypeScriptBridgeResult,
  executeTypeScriptWithGnosis,
  renderTypeScriptBridgeRuntimeModule,
  serializeTypeScriptBridgeResult,
  transpileTypeScriptBridgeRuntimeModule,
} from './ts-bridge';

describe('TypeScript to Gnosis bridge', () => {
  it('attaches selected braid fast-path metadata without changing payload', async () => {
    const source = `
      export function app(input: string) {
        return input;
      }
    `;
    const compiled = compileTypeScriptToGnosis(source, {
      exportName: 'app',
      sourceFilePath: 'braid-fast-path.ts',
    });
    const certificate = createBraidFoldCertificate({
      cycle: 1,
      gateCoordinates: [10, 12, 17, 22],
      saturationTrace: [10, 30, 90],
      closedArchitecture: true,
      endpointsLinked: true,
      baseReady: true,
    });

    const result = await executeTypeScriptWithGnosis({
      compiled,
      input: 'braid-ready',
      certifiedBraidFastPath: selectCertifiedBraidFastPath(certificate),
    });

    expect(result.payload).toBe('braid-ready');
    expect(result.braidFastPath?.status).toBe('selected');
    expect(result.braidFastPath?.hitCount).toBe(1);
    expect(result.braidFastPath?.fallbackCount).toBe(0);
    expect(result.braidFastPath?.equivalentToLinear).toBe(true);
    expect(result.braidFastPath?.runtimeDecisionCount).toBe(0);
    expect(result.braidFastPath?.avoidedRuntimeDecisionCount).toBeGreaterThan(
      0
    );
    expect(result.decision?.edgeSelectionCount).toBe(0);
    expect(result.decision?.avoidedEdgeSelectionCount).toBeGreaterThan(0);
    expect(result.braidFastPath?.collapsedBridge).toEqual({
      from: 17,
      to: 22,
      stride: 5,
    });
  });

  it('falls back when no braid fast-path certificate is supplied', async () => {
    const source = `
      export function app(input: string) {
        return input;
      }
    `;
    const compiled = compileTypeScriptToGnosis(source, {
      exportName: 'app',
      sourceFilePath: 'braid-fast-path-fallback.ts',
    });

    const result = await executeTypeScriptWithGnosis({
      compiled,
      input: 'braid-ready',
    });

    expect(result.payload).toBe('braid-ready');
    expect(result.braidFastPath?.status).toBe('fallback');
    expect(result.braidFastPath?.hitCount).toBe(0);
    expect(result.braidFastPath?.fallbackCount).toBe(1);
    expect(result.braidFastPath?.equivalentToLinear).toBe(true);
    expect(result.braidFastPath?.runtimeDecisionCount).toBeGreaterThan(0);
    expect(result.braidFastPath?.avoidedRuntimeDecisionCount).toBe(0);
    expect(result.decision?.edgeSelectionCount).toBeGreaterThan(0);
    expect(result.decision?.avoidedEdgeSelectionCount).toBe(0);
  });

  it('compiles a linear exported orchestrator into a GG process chain', async () => {
    const source = `
      export async function loadUser(userId: string) {
        return { id: userId, name: \`user-\${userId}\` };
      }

      export function summarize(user: { id: string; name: string }) {
        return user.name.toUpperCase();
      }

      export async function app(input: { userId: string }) {
        const user = await loadUser(input.userId);
        const summary = summarize(user);
        return { user, summary };
      }
    `;

    const compiled = compileTypeScriptToGnosis(source, {
      exportName: 'app',
      sourceFilePath: 'app.ts',
    });

    expect(compiled.topologySource).toContain('TsBridgeEntry');
    expect(compiled.topologySource).toContain('TsBridgeCall');
    expect(compiled.topologySource).not.toContain('[:FORK]');

    const result = await executeTypeScriptWithGnosis({
      compiled,
      input: { userId: 'ada' },
      bindings: {
        loadUser: async (userId: string) => ({
          id: userId,
          name: `user-${userId}`,
        }),
        summarize: (user: { id: string; name: string }) =>
          user.name.toUpperCase(),
      },
    });

    expect(result.payload).toEqual({
      user: { id: 'ada', name: 'user-ada' },
      summary: 'USER-ADA',
    });
  });

  it('maps Promise.all to a fork/fold topology and merges branch locals before the next step', async () => {
    const source = `
      export async function loadUser(userId: string) {
        return { id: userId, kind: 'user' };
      }

      export async function loadProfile(userId: string) {
        return { id: userId, role: 'admin' };
      }

      export function merge(
        user: { id: string; kind: string },
        profile: { id: string; role: string },
      ) {
        return \`\${user.id}:\${user.kind}:\${profile.role}\`;
      }

      export async function app(input: { userId: string }) {
        const [user, profile] = await Promise.all([
          loadUser(input.userId),
          loadProfile(input.userId),
        ]);
        const summary = merge(user, profile);
        return summary;
      }
    `;

    const compiled = compileTypeScriptToGnosis(source, {
      exportName: 'app',
      sourceFilePath: 'parallel-app.ts',
    });

    expect(compiled.topologySource).toContain('[:FORK]');
    expect(compiled.topologySource).toContain('[:FOLD]');
    expect(await checkGgProgram(compiled.topologySource)).toMatchObject({
      ok: true,
    });
    expect(compiled.schedule.map((wave) => wave.kind)).toEqual([
      'parallel',
      'collapse',
      'linear',
      'linear',
    ]);
    expect(compiled.runtimeModuleSource).toContain('@gnosis-braid-fold');
    expect(compiled.runtimeModuleSource).toContain(
      '@gnosis-decomposition-triple'
    );
    expect(compiled.runtimeModuleSource).toContain('@gnosis-braid-packet');
    expect(compiled.runtimeModuleSource).toContain(
      '@gnosis-extended-aeon-transition'
    );
    expect(compiled.runtimeModuleSource).toContain('@gnosis-interfere-port');
    expect(compiled.runtimeModuleSource).toContain(
      '@gnosis-pressure-weighted-interference'
    );
    expect(compiled.runtimeModuleSource).toContain(
      '@gnosis-closed-loop-runtime-metadata'
    );
    expect(compiled.runtimeModuleSource).toContain('"portAction":"interfere"');
    expect(compiled.runtimeModuleSource).toContain('"portCoordinate":43');
    expect(compiled.runtimeModuleSource).toContain(
      '"closedTrace":[10,12,17,22,43,55,10]'
    );
    expect(compiled.runtimeModuleSource).toContain(
      '"representativeTrace":[180,180]'
    );
    expect(compiled.runtimeModuleSource).toContain(
      '"reconstructedTrace":[360,1080]'
    );
    expect(compiled.runtimeModuleSource).toContain('"checksum":6439');
    expect(compiled.runtimeModuleSource).toContain(
      '"weightedPressureTrace":[0,558,2028]'
    );
    expect(compiled.runtimeModuleSource).toContain(
      '"weightedPressureTotal":2586'
    );
    expect(compiled.runtimeModuleSource).toContain('"coordinates":[7,17,22]');

    const result = await executeTypeScriptWithGnosis({
      compiled,
      input: { userId: 'ada' },
      bindings: {
        loadUser: async (userId: string) => ({ id: userId, kind: 'user' }),
        loadProfile: async (userId: string) => ({ id: userId, role: 'admin' }),
        merge: (
          user: { id: string; kind: string },
          profile: { id: string; role: string }
        ) => `${user.id}:${user.kind}:${profile.role}`,
      },
    });

    expect(result.payload).toBe('ada:user:admin');
  });

  it('supports member call targets and preserves the receiver', async () => {
    const source = `
      export function app(input: string) {
        console.log(input);
        return Math.max(1, 3);
      }
    `;

    const compiled = compileTypeScriptToGnosis(source, {
      exportName: 'app',
      sourceFilePath: 'member-call.ts',
    });
    const consoleStub = {
      prefix: 'captured',
      entries: [] as string[],
      log(this: { prefix: string; entries: string[] }, value: string) {
        this.entries.push(`${this.prefix}:${value}`);
      },
    };

    expect(compiled.runtimeBindingNames).toContain('Math');
    expect(compiled.runtimeBindingNames).toContain('console');

    const result = await executeTypeScriptWithGnosis({
      compiled,
      input: 'ada',
      bindings: {
        console: consoleStub,
      },
    });

    expect(result.payload).toBe(3);
    expect(consoleStub.entries).toEqual(['captured:ada']);
  });

  it('allows non-exported helper calls when the runtime binding table supplies them', async () => {
    const source = `
      function internalStep(input: number) {
        return input + 2;
      }

      export function app(input: number) {
        const value = internalStep(input);
        return value;
      }
    `;

    const compiled = compileTypeScriptToGnosis(source, {
      exportName: 'app',
      sourceFilePath: 'local-helper.ts',
    });

    expect(compiled.runtimeModuleSource).not.toContain('@gnosis-braid-fold');
    expect(compiled.runtimeModuleSource).not.toContain(
      '@gnosis-decomposition-triple'
    );
    expect(compiled.runtimeModuleSource).not.toContain('@gnosis-braid-packet');
    expect(compiled.runtimeModuleSource).not.toContain(
      '@gnosis-extended-aeon-transition'
    );
    expect(compiled.runtimeModuleSource).not.toContain(
      '@gnosis-interfere-port'
    );
    expect(compiled.runtimeModuleSource).not.toContain(
      '@gnosis-pressure-weighted-interference'
    );
    expect(compiled.runtimeModuleSource).not.toContain(
      '@gnosis-closed-loop-runtime-metadata'
    );

    const result = await executeTypeScriptWithGnosis({
      compiled,
      input: 5,
      bindings: {
        internalStep: (input: number) => input + 2,
      },
    });

    expect(result.payload).toBe(7);
  });

  it('autoloads top-level helper bindings from a module path for gnode-style execution', async () => {
    const tempDirectory = fs.mkdtempSync(
      path.join(os.tmpdir(), 'gnode-bridge-')
    );
    const modulePath = path.join(tempDirectory, 'app.ts');

    fs.writeFileSync(
      modulePath,
      `
        function internalStep(input: number) {
          return input + 2;
        }

        export function app(input: number) {
          const value = internalStep(input);
          return value;
        }
      `,
      'utf8'
    );

    try {
      const result = await executeTypeScriptWithGnosis({
        modulePath,
        exportName: 'app',
        input: 5,
      });

      expect(result.payload).toBe(7);
    } finally {
      fs.rmSync(tempDirectory, { recursive: true, force: true });
    }
  });

  it('hydrates a serialized compiled artifact without reparsing the GG topology', async () => {
    const source = `
      function internalStep(input: number) {
        return input + 2;
      }

      export function app(input: number) {
        const value = internalStep(input);
        return value;
      }
    `;

    const compiled = compileTypeScriptToGnosis(source, {
      exportName: 'app',
      sourceFilePath: 'serialized-app.ts',
    });
    const hydrated = deserializeTypeScriptBridgeResult(
      serializeTypeScriptBridgeResult(compiled)
    );

    expect(hydrated.runtimeBindingNames).toEqual(compiled.runtimeBindingNames);
    expect(hydrated.runtimeModuleSource).toBe(compiled.runtimeModuleSource);

    const result = await executeTypeScriptWithGnosis({
      compiled: hydrated,
      input: 5,
      bindings: {
        internalStep: (input: number) => input + 2,
      },
    });

    expect(result.payload).toBe(7);
  });

  it('executes against a cached runtime module outside the source tree', async () => {
    const sourceDirectory = fs.mkdtempSync(
      path.join(os.tmpdir(), 'gnode-source-')
    );
    const runtimeDirectory = fs.mkdtempSync(
      path.join(os.tmpdir(), 'gnode-runtime-')
    );
    const modulePath = path.join(sourceDirectory, 'app.ts');
    const helperPath = path.join(sourceDirectory, 'helper.ts');
    const runtimeModulePath = path.join(runtimeDirectory, 'runtime-bridge.ts');

    fs.writeFileSync(
      helperPath,
      `
        export function internalStep(input: number) {
          return input + 2;
        }
      `,
      'utf8'
    );
    fs.writeFileSync(
      modulePath,
      `
        import { internalStep } from './helper';

        export function app(input: number) {
          const value = internalStep(input);
          return value;
        }
      `,
      'utf8'
    );

    try {
      const source = fs.readFileSync(modulePath, 'utf8');
      const compiled = compileTypeScriptToGnosis(source, {
        exportName: 'app',
        sourceFilePath: modulePath,
      });
      const runtimeModule = renderTypeScriptBridgeRuntimeModule(
        source,
        modulePath,
        compiled.runtimeBindingNames,
        { specifierStyle: 'absolute-url' }
      );

      fs.writeFileSync(runtimeModulePath, runtimeModule.moduleSource, 'utf8');

      const result = await executeTypeScriptWithGnosis({
        compiled,
        runtimeModulePath,
        input: 5,
      });

      expect(result.payload).toBe(7);
    } finally {
      fs.rmSync(sourceDirectory, { recursive: true, force: true });
      fs.rmSync(runtimeDirectory, { recursive: true, force: true });
    }
  });

  it('rewrites explicit .ts imports when a cached runtime module executes outside the source tree', async () => {
    const sourceDirectory = fs.mkdtempSync(
      path.join(os.tmpdir(), 'gnode-source-ts-')
    );
    const runtimeDirectory = fs.mkdtempSync(
      path.join(os.tmpdir(), 'gnode-runtime-ts-')
    );
    const modulePath = path.join(sourceDirectory, 'app.ts');
    const helperPath = path.join(sourceDirectory, 'helper.ts');
    const runtimeModulePath = path.join(runtimeDirectory, 'runtime-bridge.mjs');

    fs.writeFileSync(
      helperPath,
      `
        export function internalStep(input: number) {
          return input + 3;
        }
      `,
      'utf8'
    );
    fs.writeFileSync(
      modulePath,
      `
        import { internalStep } from './helper';

        export function app(input: number) {
          const value = internalStep(input);
          return value;
        }
      `,
      'utf8'
    );

    try {
      const source = fs.readFileSync(modulePath, 'utf8');
      const compiled = compileTypeScriptToGnosis(source, {
        exportName: 'app',
        sourceFilePath: modulePath,
      });
      const runtimeModule = transpileTypeScriptBridgeRuntimeModule(
        source,
        modulePath,
        compiled.runtimeBindingNames,
        { specifierStyle: 'absolute-url' }
      );

      expect(runtimeModule.javascriptSource).toContain(
        pathToFileURL(helperPath).href
      );

      fs.writeFileSync(
        runtimeModulePath,
        runtimeModule.javascriptSource,
        'utf8'
      );

      const result = await executeTypeScriptWithGnosis({
        compiled,
        runtimeModulePath,
        input: 5,
      });

      expect(result.payload).toBe(8);
    } finally {
      fs.rmSync(sourceDirectory, { recursive: true, force: true });
      fs.rmSync(runtimeDirectory, { recursive: true, force: true });
    }
  });

  it('ignores type-only imports when rendering runtime bridge bindings', () => {
    const source = `
      import { spawn, type ChildProcess } from 'node:child_process';

      function launch() {
        return spawn;
      }

      export function app() {
        return launch();
      }
    `;

    const compiled = compileTypeScriptToGnosis(source, {
      exportName: 'app',
      sourceFilePath: 'type-only-import.ts',
    });
    const runtimeModule = renderTypeScriptBridgeRuntimeModule(
      source,
      'type-only-import.ts',
      compiled.runtimeBindingNames
    );

    expect(compiled.runtimeBindingNames).toContain('spawn');
    expect(compiled.runtimeBindingNames).not.toContain('ChildProcess');
    expect(runtimeModule.moduleSource).toContain('spawn');
    expect(runtimeModule.moduleSource).not.toContain(
      '"ChildProcess": ChildProcess'
    );
  });

  it('includes top-level const dependencies for helper runtime bindings', async () => {
    const tempDirectory = fs.mkdtempSync(
      path.join(os.tmpdir(), 'gnode-bridge-const-deps-')
    );
    const modulePath = path.join(tempDirectory, 'app.ts');

    fs.writeFileSync(
      modulePath,
      `
        const prefix = 'bridge-';

        function decorate(input: string) {
          return \`\${prefix}\${input}\`;
        }

        export function app(input: string) {
          return decorate(input);
        }
      `,
      'utf8'
    );

    try {
      const result = await executeTypeScriptWithGnosis({
        modulePath,
        exportName: 'app',
        input: 'native',
      });

      expect(result.payload).toBe('bridge-native');
    } finally {
      fs.rmSync(tempDirectory, { recursive: true, force: true });
    }
  });

  it('loads the bridge under a tsx eval entrypoint without treating [eval] as a require base', () => {
    const tempDirectory = fs.mkdtempSync(
      path.join(os.tmpdir(), 'gnode-ts-bridge-eval-')
    );
    const tsxLoaderPath = path.resolve(
      __dirname,
      '../../../node_modules/tsx/dist/loader.mjs'
    );
    const bridgeModuleUrl = pathToFileURL(
      path.resolve(__dirname, './ts-bridge.ts')
    ).href;

    try {
      const result = spawnSync(
        'node',
        [
          '--import',
          tsxLoaderPath,
          '--eval',
          `import(${JSON.stringify(
            bridgeModuleUrl
          )}).then(() => console.log('bridge-ok')).catch((error) => { console.error(error); process.exit(1); })`,
        ],
        {
          cwd: tempDirectory,
          encoding: 'utf8',
          env: {
            ...process.env,
            HOME: path.join(tempDirectory, 'home'),
          },
        }
      );

      expect(result.status).toBe(0);
      expect(result.stderr).not.toContain('ERR_INVALID_ARG_VALUE');
      expect(result.stderr).not.toContain('[eval]');
    } finally {
      fs.rmSync(tempDirectory, { recursive: true, force: true });
    }
  });

  it('rewrites explicit .js imports back to source modules for cached runtime execution', async () => {
    const sourceDirectory = fs.mkdtempSync(
      path.join(os.tmpdir(), 'gnode-source-js-')
    );
    const runtimeDirectory = fs.mkdtempSync(
      path.join(os.tmpdir(), 'gnode-runtime-js-')
    );
    const modulePath = path.join(sourceDirectory, 'app.ts');
    const helperPath = path.join(sourceDirectory, 'helper.ts');
    const runtimeModulePath = path.join(runtimeDirectory, 'runtime-bridge.mjs');

    fs.writeFileSync(
      helperPath,
      `
        export function internalStep(input: number) {
          return input + 4;
        }
      `,
      'utf8'
    );
    fs.writeFileSync(
      modulePath,
      `
        import { internalStep } from './helper';

        export function app(input: number) {
          const value = internalStep(input);
          return value;
        }
      `,
      'utf8'
    );

    try {
      const source = fs.readFileSync(modulePath, 'utf8');
      const compiled = compileTypeScriptToGnosis(source, {
        exportName: 'app',
        sourceFilePath: modulePath,
      });
      const runtimeModule = transpileTypeScriptBridgeRuntimeModule(
        source,
        modulePath,
        compiled.runtimeBindingNames,
        { specifierStyle: 'absolute-url' }
      );

      expect(runtimeModule.javascriptSource).toContain(
        pathToFileURL(helperPath).href
      );

      fs.writeFileSync(
        runtimeModulePath,
        runtimeModule.javascriptSource,
        'utf8'
      );

      const result = await executeTypeScriptWithGnosis({
        compiled,
        runtimeModulePath,
        input: 5,
      });

      expect(result.payload).toBe(9);
    } finally {
      fs.rmSync(sourceDirectory, { recursive: true, force: true });
      fs.rmSync(runtimeDirectory, { recursive: true, force: true });
    }
  });

  it('allows callers to reuse the runtime core cache across bridge executions', async () => {
    const source = `
      export function double(input: number) {
        return input * 2;
      }

      export function app(input: number) {
        const value = double(input);
        return value;
      }
    `;

    const compiled = compileTypeScriptToGnosis(source, {
      exportName: 'app',
      sourceFilePath: 'cached-app.ts',
    });
    const coreCache = new GnosisCoreCache(
      new QDoc({ guid: 'ts-bridge-cache-test' })
    );

    const first = await executeTypeScriptWithGnosis({
      compiled,
      input: 9,
      bindings: {
        double: (input: number) => input * 2,
      },
      engineOptions: {
        coreCache,
      },
      executeOptions: {
        cache: {
          corridorKey: 'x-gnosis/gnode/test',
          reuseScope: 'corridor',
        },
      },
    });

    const second = await executeTypeScriptWithGnosis({
      compiled,
      input: 11,
      bindings: {
        double: (input: number) => input * 2,
      },
      engineOptions: {
        coreCache,
      },
      executeOptions: {
        cache: {
          corridorKey: 'x-gnosis/gnode/test',
          reuseScope: 'corridor',
        },
      },
    });

    expect(first.cache?.status).toBe('miss');
    expect(second.cache?.status).toBe('hit');
    expect(second.payload).toBe(18);
  });

  it('fails on unsupported control-flow that cannot be represented honestly yet', () => {
    const source = `
      export function app(input: { enabled: boolean }) {
        if (input.enabled) {
          return 'on';
        }
        return 'off';
      }
    `;

    expect(() =>
      compileTypeScriptToGnosis(source, {
        exportName: 'app',
        sourceFilePath: 'unsupported.ts',
      })
    ).toThrow('Unsupported TypeScript statement');
  });
});
