// @a0n/.ignored_gnosis/src/ts-bridge.lift-extensions.test.ts
// ts-bridge lift-extensions self-test
//
// Verifies that the gnode bridge lifts two new control-flow patterns:
//   1. `for (const x of [a, b, c]) { x = await call(...) }` -> PROCESS chain
//      (statically unrolled into N call nodes connected by PROCESS edges).
//   2. `await Promise.race([call(...), call(...)])` -> RACE node (FORK to
//      branches + RACE-collapse to a TsBridgeRace resolver node).
//
// Run directly:
//   pnpm gnode run open-source/gnosis/src/ts-bridge.lift-extensions.test.ts
// or:
//   npx tsx open-source/gnosis/src/ts-bridge.lift-extensions.test.ts
//
// Exits with code 0 on success, 1 on first failure. No external test runner
// is required (deliberately -- the task scope forbids adding test infra).

import { compileTypeScriptToGnosis } from './ts-bridge';

let failures = 0;

function check(label: string, condition: boolean, detail?: string): void {
  const status = condition ? 'PASS' : 'FAIL';
   
  console.log(`  [${status}] ${label}${condition || !detail ? '' : ` -- ${detail}`}`);
  if (!condition) {
    failures += 1;
  }
}

// ── Test 1: for-of PROCESS chain ──────────────────────────────────────────────
{
   
  console.log('for-of PROCESS chain (3 iterations)');
  const source = `
    export async function run(input: number): Promise<number> {
      let x = input;
      for (const layer of [0, 1, 2]) {
        x = await call(layer, x);
      }
      return x;
    }
  `;
  const compiled = compileTypeScriptToGnosis(source, {
    exportName: 'run',
    sourceFilePath: 'lift-extensions-for-of.ts',
  });

  // Count TsBridgeCall nodes — should be 3 (one per loop iteration).
  const callPlans = compiled.nodePlans.filter(
    (plan) => plan.handlerLabel === 'TsBridgeCall'
  );
  check(
    'emits one TsBridgeCall per unrolled iteration',
    callPlans.length === 3,
    `expected 3, got ${callPlans.length}`
  );

  // Count assign plans for loop var `layer` — one per iteration.
  const layerAssigns = compiled.nodePlans.filter(
    (plan) => plan.handlerLabel === 'TsBridgeAssign' &&
      // assign plans carry a bindingName field
      (plan as { bindingName?: string }).bindingName === 'layer'
  );
  check(
    'emits one assign for loop variable per iteration',
    layerAssigns.length === 3,
    `expected 3, got ${layerAssigns.length}`
  );

  // PROCESS edges should connect calls in a chain. We expect at least 6
  // PROCESS edges (entry -> assign -> call alternating, plus a return).
  const processEdges = compiled.ast.edges.filter(
    (edge) => edge.type === 'PROCESS'
  );
  check(
    'emits PROCESS edges threading the chain',
    processEdges.length >= 6,
    `expected >= 6 PROCESS edges, got ${processEdges.length}`
  );

  // No FORK edges should appear for a pure for-of chain.
  const forkEdges = compiled.ast.edges.filter((edge) => edge.type === 'FORK');
  check(
    'no FORK edges in a pure for-of chain',
    forkEdges.length === 0,
    `unexpected FORK edges: ${forkEdges.length}`
  );

  // Total node count check: for trisplit-llm-gg.ts comparison, we want to
  // know how many nodes a 3-iteration for-of contributes. Report it.
   
  console.log(
    `  (info) total node plans for 3-iteration for-of: ${compiled.nodePlans.length}`
  );
}

// ── Test 2: Promise.race RACE lifter ──────────────────────────────────────────
{
   
  console.log('Promise.race RACE lifter');
  const source = `
    export async function pick(input: number): Promise<number> {
      const winner = await Promise.race([
        call(input, 'a'),
        call(input, 'b'),
      ]);
      return winner;
    }
  `;
  const compiled = compileTypeScriptToGnosis(source, {
    exportName: 'pick',
    sourceFilePath: 'lift-extensions-race.ts',
  });

  // One TsBridgeRace node should be present.
  const racePlans = compiled.nodePlans.filter(
    (plan) => plan.handlerLabel === 'TsBridgeRace'
  );
  check(
    'emits exactly one TsBridgeRace node',
    racePlans.length === 1,
    `expected 1, got ${racePlans.length}`
  );

  // Two TsBridgeCall branches.
  const callPlans = compiled.nodePlans.filter(
    (plan) => plan.handlerLabel === 'TsBridgeCall'
  );
  check(
    'emits one TsBridgeCall per race branch',
    callPlans.length === 2,
    `expected 2, got ${callPlans.length}`
  );

  // One FORK edge from current -> branches.
  const forkEdges = compiled.ast.edges.filter((edge) => edge.type === 'FORK');
  check(
    'emits FORK edge to race branches',
    forkEdges.length === 1,
    `expected 1, got ${forkEdges.length}`
  );

  // One RACE edge from branches -> resolver.
  const raceEdges = compiled.ast.edges.filter((edge) => edge.type === 'RACE');
  check(
    'emits RACE edge from branches to resolver',
    raceEdges.length === 1,
    `expected 1, got ${raceEdges.length}`
  );

  // No FOLD edge (RACE is distinct from FORK/FOLD).
  const foldEdges = compiled.ast.edges.filter((edge) => edge.type === 'FOLD');
  check(
    'no FOLD edge in a pure race',
    foldEdges.length === 0,
    `unexpected FOLD edges: ${foldEdges.length}`
  );

  // Topology source should mention `[:RACE]`.
  check(
    'topology source contains [:RACE] edge',
    compiled.topologySource.includes('[:RACE]'),
    'topologySource missing [:RACE] label'
  );
}

// ── Test 3: backward compatibility — Promise.all still emits FORK/FOLD ────────
{
   
  console.log('backward compatibility: Promise.all -> FORK + FOLD');
  const source = `
    export async function fan(input: number): Promise<number[]> {
      const [a, b] = await Promise.all([
        call(input, 'a'),
        call(input, 'b'),
      ]);
      return [a, b];
    }
  `;
  const compiled = compileTypeScriptToGnosis(source, {
    exportName: 'fan',
    sourceFilePath: 'lift-extensions-promise-all.ts',
  });

  const forkEdges = compiled.ast.edges.filter((edge) => edge.type === 'FORK');
  const foldEdges = compiled.ast.edges.filter((edge) => edge.type === 'FOLD');
  const raceEdges = compiled.ast.edges.filter((edge) => edge.type === 'RACE');

  check('Promise.all still emits FORK', forkEdges.length === 1);
  check('Promise.all still emits FOLD', foldEdges.length === 1);
  check('Promise.all does not emit RACE', raceEdges.length === 0);
}

 
console.log('');
if (failures === 0) {
   
  console.log('ts-bridge lift-extensions: ALL CHECKS PASSED');
  process.exit(0);
} else {
   
  console.error(`ts-bridge lift-extensions: ${failures} CHECK(S) FAILED`);
  process.exit(1);
}
