// @a0n/.ignored_gnosis/src/ts-check-autofix.ts
/**
 * Gnosis Topology Autofix Engine
 *
 * Analyzes topology violations and generates concrete code fixes:
 * - Split functions along FORK boundaries when buley exceeds threshold
 * - Insert FOLD (Promise.race / error boundaries) for uncollapsed superposition
 * - Remove dead branches (unreachable nodes)
 * - Suggest restructuring for turbulent steering regimes
 */

import type {
  GnosisTypeScriptCheckResult,
//   _GnosisTypeScriptCheckDiagnostic,
  GnosisTypeScriptCheckTopologyNode,
  GnosisTypeScriptCheckTopologyEdge,
} from './ts-check';

export interface AutofixSuggestion {
  readonly ruleId: string;
  readonly kind: AutofixKind;
  readonly message: string;
  readonly line: number;
  readonly column: number;
  readonly endLine?: number;
  readonly endColumn?: number;
  readonly replacement?: AutofixReplacement;
  readonly severity: 'info' | 'suggestion' | 'refactor';
}

export type AutofixKind =
  | 'split-function'
  | 'insert-fold'
  | 'remove-dead-branch'
  | 'extract-parallel'
  | 'add-error-boundary'
  | 'reduce-nesting';

export interface AutofixReplacement {
  readonly startLine: number;
  readonly startColumn: number;
  readonly endLine: number;
  readonly endColumn: number;
  readonly newText: string;
}

export interface DeadBranchInfo {
  readonly nodeId: string;
  readonly kind: string;
  readonly sourceLocation?: {
    readonly line: number;
    readonly column: number;
    readonly endLine?: number;
    readonly endColumn?: number;
  };
  readonly reason: string;
}

// ---------------------------------------------------------------------------
// Dead branch detection
// ---------------------------------------------------------------------------

/**
 * Handles the gnosis detect Dead Branches workflow.
 */
export function detectDeadBranches(
  result: GnosisTypeScriptCheckResult
): readonly DeadBranchInfo[] {
  const { nodes, edges } = result.topology;
  if (nodes.length === 0) return [];

  // Find entry node
  const entryNode = nodes.find((n) => n.kind === 'entry');
  if (!entryNode) return [];

  // BFS from entry to find reachable nodes
  const adjacency = new Map<string, string[]>();
  for (const edge of edges) {
    const targets = adjacency.get(edge.from) ?? [];
    targets.push(edge.to);
    adjacency.set(edge.from, targets);
  }

  const reachable = new Set<string>();
  const queue = [entryNode.id];
  while (queue.length > 0) {
    const current = queue.shift()!;
    if (reachable.has(current)) continue;
    reachable.add(current);
    for (const target of adjacency.get(current) ?? []) {
      if (!reachable.has(target)) {
        queue.push(target);
      }
    }
  }

  // Unreachable nodes are dead branches
  return nodes
    .filter((n) => !reachable.has(n.id))
    .map((n) => ({
      nodeId: n.id,
      kind: n.kind,
      reason: 'Node is unreachable from entry point',
      ...(n.sourceLocation !== undefined
        ? { sourceLocation: n.sourceLocation }
        : {}),
    }));
}

// ---------------------------------------------------------------------------
// FORK boundary analysis
// ---------------------------------------------------------------------------

interface ForkBoundary {
  readonly forkEdge: GnosisTypeScriptCheckTopologyEdge;
  readonly branches: readonly GnosisTypeScriptCheckTopologyNode[];
  readonly foldEdge: GnosisTypeScriptCheckTopologyEdge | null;
  readonly branchComplexity: number;
}

function findForkBoundaries(
  result: GnosisTypeScriptCheckResult
): readonly ForkBoundary[] {
  const { nodes, edges } = result.topology;
  const nodeMap = new Map(nodes.map((n) => [n.id, n]));
  const boundaries: ForkBoundary[] = [];

  const forkEdges = edges.filter((e) => e.type === 'FORK');
  for (const forkEdge of forkEdges) {
    const branches: GnosisTypeScriptCheckTopologyNode[] = [];
    // Find all nodes that this fork leads to
    for (const edge of edges) {
      if (edge.from === forkEdge.from && edge.type === 'FORK') {
        const target = nodeMap.get(edge.to);
        if (target) branches.push(target);
      }
    }

    // Find corresponding FOLD
    const foldEdge =
      edges.find(
        (e) => e.type === 'FOLD' && branches.some((b) => b.id === e.from)
      ) ?? null;

    boundaries.push({
      forkEdge,
      branches,
      foldEdge,
      branchComplexity: branches.length,
    });
  }

  return boundaries;
}

// ---------------------------------------------------------------------------
// Autofix generation
// ---------------------------------------------------------------------------

/**
 * Handles the gnosis generate Autofix Suggestions workflow.
 */
export function generateAutofixSuggestions(
  result: GnosisTypeScriptCheckResult,
  _sourceText: string
): readonly AutofixSuggestion[] {
  const suggestions: AutofixSuggestion[] = [];

  // 1. Dead branch removal
  const deadBranches = detectDeadBranches(result);
  for (const dead of deadBranches) {
    if (dead.sourceLocation) {
      const sourceLocation = dead.sourceLocation;
      suggestions.push({
        ruleId: 'gnosis/dead-branch',
        kind: 'remove-dead-branch',
        message: `Dead branch: ${dead.kind} node '${dead.nodeId}' is unreachable from entry. Consider removing the associated code.`,
        line: sourceLocation.line,
        column: sourceLocation.column,
        ...(sourceLocation.endLine !== undefined
          ? { endLine: sourceLocation.endLine }
          : {}),
        ...(sourceLocation.endColumn !== undefined
          ? { endColumn: sourceLocation.endColumn }
          : {}),
        severity: 'suggestion',
      });
    }
  }

  // 2. Buley threshold split suggestions
  if (result.metrics.buleyNumber > 50) {
    const forkBoundaries = findForkBoundaries(result);
    if (forkBoundaries.length > 0) {
      for (const boundary of forkBoundaries) {
        const branchNames = boundary.branches.map((b) => b.label).join(', ');
        suggestions.push({
          ruleId: 'gnosis/split-at-fork',
          kind: 'split-function',
          message: `High complexity (buley=${result.metrics.buleyNumber}). Consider extracting branches [${branchNames}] into separate functions. The FORK boundary is a natural split point.`,
          line: boundary.branches[0]?.sourceLocation?.line ?? 1,
          column: boundary.branches[0]?.sourceLocation?.column ?? 1,
          severity: 'refactor',
        });
      }
    } else {
      // No FORK boundaries, suggest general splitting
      const midpoint = Math.floor(result.topology.nodes.length / 2);
      const midNode = result.topology.nodes[midpoint];
      suggestions.push({
        ruleId: 'gnosis/split-function',
        kind: 'split-function',
        message: `High complexity (buley=${result.metrics.buleyNumber}). Consider splitting this function into smaller units. The topology has ${result.topology.nodes.length} nodes — aim for functions with buley < 30.`,
        line: midNode?.sourceLocation?.line ?? 1,
        column: midNode?.sourceLocation?.column ?? 1,
        severity: 'refactor',
      });
    }
  }

  // 3. Uncollapsed superposition (FORK without matching FOLD)
  const forkEdges = result.topology.edges.filter((e) => e.type === 'FORK');
  const foldEdges = result.topology.edges.filter((e) => e.type === 'FOLD');
  if (forkEdges.length > foldEdges.length) {
    const deficit = forkEdges.length - foldEdges.length;
    suggestions.push({
      ruleId: 'gnosis/uncollapsed-superposition',
      kind: 'insert-fold',
      message: `${deficit} FORK edge(s) without matching FOLD. Each parallel branch should be collapsed with Promise.all, Promise.race, or an error boundary.`,
      line: 1,
      column: 1,
      severity: 'suggestion',
    });
  }

  // 4. Turbulent regime suggestions
  if (result.metrics.regime === 'turbulent') {
    suggestions.push({
      ruleId: 'gnosis/turbulent-regime',
      kind: 'reduce-nesting',
      message: `Steering regime is turbulent (wallace=${result.metrics.wallaceNumber}). Consider: (1) reducing parallel branch depth, (2) adding intermediate collapse points, (3) splitting into pipeline stages.`,
      line: 1,
      column: 1,
      severity: 'refactor',
    });
  }

  // 5. Sequential chains that could be parallelized
  const processChains = findLongProcessChains(result);
  for (const chain of processChains) {
    if (chain.length >= 3) {
      const independentPairs = findIndependentNodes(chain, result);
      if (independentPairs.length > 0) {
        const firstNode = chain[0];
        if (firstNode === undefined) {
          continue;
        }
        suggestions.push({
          ruleId: 'gnosis/parallelize-chain',
          kind: 'extract-parallel',
          message: `Sequential chain of ${chain.length} operations could be partially parallelized. ${independentPairs.length} pair(s) of independent operations detected. Consider wrapping in Promise.all.`,
          line: firstNode.sourceLocation?.line ?? 1,
          column: firstNode.sourceLocation?.column ?? 1,
          severity: 'info',
        });
      }
    }
  }

  return suggestions;
}

function findLongProcessChains(
  result: GnosisTypeScriptCheckResult
): readonly (readonly GnosisTypeScriptCheckTopologyNode[])[] {
  const { nodes, edges } = result.topology;
  const nodeMap = new Map(nodes.map((n) => [n.id, n]));
  const processEdges = edges.filter((e) => e.type === 'PROCESS');

  // Build adjacency for PROCESS edges only
  const next = new Map<string, string>();
  const prev = new Map<string, string>();
  for (const edge of processEdges) {
    next.set(edge.from, edge.to);
    prev.set(edge.to, edge.from);
  }

  // Find chain starts (nodes with no PROCESS predecessor)
  const chainStarts = nodes.filter((n) => !prev.has(n.id) && next.has(n.id));

  const chains: GnosisTypeScriptCheckTopologyNode[][] = [];
  for (const start of chainStarts) {
    const chain: GnosisTypeScriptCheckTopologyNode[] = [start];
    let current = start.id;
    while (next.has(current)) {
      const nextId = next.get(current)!;
      const nextNode = nodeMap.get(nextId);
      if (!nextNode) break;
      chain.push(nextNode);
      current = nextId;
    }
    if (chain.length >= 3) {
      chains.push(chain);
    }
  }

  return chains;
}

function findIndependentNodes(
  chain: readonly GnosisTypeScriptCheckTopologyNode[],
  _result: GnosisTypeScriptCheckResult
): readonly [
  GnosisTypeScriptCheckTopologyNode,
  GnosisTypeScriptCheckTopologyNode
][] {
  // Two nodes are independent if neither reads from the other's output
  // Heuristic: call nodes with different callee names and no shared bindings
  const pairs: [
    GnosisTypeScriptCheckTopologyNode,
    GnosisTypeScriptCheckTopologyNode
  ][] = [];

  for (let i = 0; i < chain.length; i++) {
    for (let j = i + 1; j < chain.length; j++) {
      const a = chain[i];
      const b = chain[j];
      if (a === undefined || b === undefined) {
        continue;
      }
      if (a.kind === 'call' && b.kind === 'call' && a.label !== b.label) {
        // Different function calls with no data dependency (heuristic)
        pairs.push([a, b]);
      }
    }
  }

  return pairs;
}
