// @a0n/.ignored_gnosis/src/ts-check-diff.ts
/**
 * Gnosis Topology Diff
 *
 * Computes the delta in topology metrics between two versions of a file
 * (e.g., main branch vs PR branch). Used by CI pipelines and PR review bots.
 */

import * as path from 'node:path';
import type {
  GnosisTypeScriptCheckMetrics,
  GnosisTypeScriptCheckResult,
//   _GnosisTypeScriptCheckTopologyNode,
} from './ts-check';
import {
  checkTypeScriptWithGnosis,
  type GnosisTypeScriptCheckOptions,
} from './ts-check';
import type { GnosisSteeringRegime } from './analysis';

export interface TopologyMetricsDelta {
  readonly buleyNumber: number;
  readonly wallaceNumber: number;
  readonly beta1: number;
  readonly quantumIndex: number;
  readonly nodeCount: number;
  readonly edgeCount: number;
}

export interface TopologyRegimeTransition {
  readonly from: GnosisSteeringRegime;
  readonly to: GnosisSteeringRegime;
  readonly degraded: boolean;
  readonly improved: boolean;
}

export interface FunctionDiff {
  readonly functionName: string;
  readonly filePath: string;
  readonly status: 'added' | 'removed' | 'changed' | 'unchanged';
  readonly baseBuley: number | null;
  readonly headBuley: number | null;
  readonly delta: TopologyMetricsDelta | null;
  readonly regimeTransition: TopologyRegimeTransition | null;
  readonly baseMetrics: GnosisTypeScriptCheckMetrics | null;
  readonly headMetrics: GnosisTypeScriptCheckMetrics | null;
}

export interface TopologyDiffResult {
  readonly filePath: string;
  readonly functions: readonly FunctionDiff[];
  readonly aggregateDelta: TopologyMetricsDelta;
  readonly worstRegimeTransition: TopologyRegimeTransition | null;
  readonly regressions: readonly string[];
  readonly improvements: readonly string[];
  readonly summary: string;
}

export interface TopologyDiffOptions extends GnosisTypeScriptCheckOptions {
  readonly buleyRegressionThreshold?: number;
}

const REGIME_ORDER: Record<GnosisSteeringRegime, number> = {
  laminar: 0,
  transitional: 1,
  turbulent: 2,
};

/**
 * Computes the Topology Diff.
 */
export async function computeTopologyDiff(
  baseSource: string,
  headSource: string,
  filePath: string,
  options: TopologyDiffOptions = {}
): Promise<TopologyDiffResult> {
  const threshold = options.buleyRegressionThreshold ?? 5;

  let baseResult: GnosisTypeScriptCheckResult | null = null;
  let headResult: GnosisTypeScriptCheckResult | null = null;

  try {
    baseResult = await checkTypeScriptWithGnosis(baseSource, filePath, options);
  } catch {
    // Base may not have bridgeable functions
  }

  try {
    headResult = await checkTypeScriptWithGnosis(headSource, filePath, options);
  } catch {
    // Head may not have bridgeable functions
  }

  const functions: FunctionDiff[] = [];
  const regressions: string[] = [];
  const improvements: string[] = [];

  if (!baseResult && headResult) {
    // New function
    functions.push({
      functionName: 'default',
      filePath,
      status: 'added',
      baseBuley: null,
      headBuley: headResult.metrics.buleyNumber,
      delta: null,
      regimeTransition: null,
      baseMetrics: null,
      headMetrics: headResult.metrics,
    });
  } else if (baseResult && !headResult) {
    // Removed function
    functions.push({
      functionName: 'default',
      filePath,
      status: 'removed',
      baseBuley: baseResult.metrics.buleyNumber,
      headBuley: null,
      delta: null,
      regimeTransition: null,
      baseMetrics: baseResult.metrics,
      headMetrics: null,
    });
    improvements.push(
      `Removed function (was buley=${baseResult.metrics.buleyNumber})`
    );
  } else if (baseResult && headResult) {
    const delta = computeDelta(baseResult.metrics, headResult.metrics);
    const transition = computeRegimeTransition(
      baseResult.metrics.regime,
      headResult.metrics.regime
    );

    const status =
      delta.buleyNumber === 0 &&
      delta.wallaceNumber === 0 &&
      delta.nodeCount === 0
        ? 'unchanged'
        : 'changed';

    functions.push({
      functionName: 'default',
      filePath,
      status,
      baseBuley: baseResult.metrics.buleyNumber,
      headBuley: headResult.metrics.buleyNumber,
      delta,
      regimeTransition: transition,
      baseMetrics: baseResult.metrics,
      headMetrics: headResult.metrics,
    });

    if (delta.buleyNumber > threshold) {
      regressions.push(
        `Buley increased by ${delta.buleyNumber} (${baseResult.metrics.buleyNumber} -> ${headResult.metrics.buleyNumber})`
      );
    }
    if (transition.degraded) {
      regressions.push(
        `Regime degraded: ${transition.from} -> ${transition.to}`
      );
    }
    if (delta.buleyNumber < -threshold) {
      improvements.push(
        `Buley decreased by ${Math.abs(delta.buleyNumber)} (${
          baseResult.metrics.buleyNumber
        } -> ${headResult.metrics.buleyNumber})`
      );
    }
    if (transition.improved) {
      improvements.push(
        `Regime improved: ${transition.from} -> ${transition.to}`
      );
    }
  }

  const aggregateDelta = functions[0]?.delta ?? zeroDelta();
  const worstTransition =
    functions
      .map((f) => f.regimeTransition)
      .filter((t): t is TopologyRegimeTransition => t !== null)
      .sort((a, b) => {
        if (a.degraded && !b.degraded) return -1;
        if (!a.degraded && b.degraded) return 1;
        return 0;
      })[0] ?? null;

  const summary = buildSummary(functions, regressions, improvements);

  return {
    filePath,
    functions,
    aggregateDelta,
    worstRegimeTransition: worstTransition,
    regressions,
    improvements,
    summary,
  };
}

function computeDelta(
  base: GnosisTypeScriptCheckMetrics,
  head: GnosisTypeScriptCheckMetrics
): TopologyMetricsDelta {
  return {
    buleyNumber: round2(head.buleyNumber - base.buleyNumber),
    wallaceNumber: round2(head.wallaceNumber - base.wallaceNumber),
    beta1: round2(head.beta1 - base.beta1),
    quantumIndex: round2(head.quantumIndex - base.quantumIndex),
    nodeCount: head.nodeCount - base.nodeCount,
    edgeCount: head.edgeCount - base.edgeCount,
  };
}

function computeRegimeTransition(
  from: GnosisSteeringRegime,
  to: GnosisSteeringRegime
): TopologyRegimeTransition {
  return {
    from,
    to,
    degraded: REGIME_ORDER[to] > REGIME_ORDER[from],
    improved: REGIME_ORDER[to] < REGIME_ORDER[from],
  };
}

function zeroDelta(): TopologyMetricsDelta {
  return {
    buleyNumber: 0,
    wallaceNumber: 0,
    beta1: 0,
    quantumIndex: 0,
    nodeCount: 0,
    edgeCount: 0,
  };
}

function round2(value: number): number {
  return Math.round(value * 100) / 100;
}

function buildSummary(
  functions: readonly FunctionDiff[],
  regressions: readonly string[],
  improvements: readonly string[]
): string {
  const lines: string[] = [];

  for (const func of functions) {
    const arrow = func.delta
      ? `${func.baseBuley} -> ${func.headBuley} (${
          func.delta.buleyNumber >= 0 ? '+' : ''
        }${func.delta.buleyNumber})`
      : func.status === 'added'
      ? `new (buley=${func.headBuley})`
      : `removed (was buley=${func.baseBuley})`;

    const regime = func.regimeTransition
      ? ` [${func.regimeTransition.from} -> ${func.regimeTransition.to}]`
      : func.headMetrics
      ? ` [${func.headMetrics.regime}]`
      : '';

    lines.push(`${func.status}: ${func.filePath} buley: ${arrow}${regime}`);
  }

  if (regressions.length > 0) {
    lines.push('');
    lines.push('Regressions:');
    for (const r of regressions) {
      lines.push(`  - ${r}`);
    }
  }

  if (improvements.length > 0) {
    lines.push('');
    lines.push('Improvements:');
    for (const imp of improvements) {
      lines.push(`  - ${imp}`);
    }
  }

  return lines.join('\n');
}

// ---------------------------------------------------------------------------
// PR comment formatting
// ---------------------------------------------------------------------------

/**
 * Formats the Pr Comment.
 */
export function formatPrComment(diffs: readonly TopologyDiffResult[]): string {
  const lines: string[] = ['## Gnosis Topology Diff', ''];

  const totalRegressions = diffs.flatMap((d) => d.regressions);
  const totalImprovements = diffs.flatMap((d) => d.improvements);

  if (totalRegressions.length === 0 && totalImprovements.length === 0) {
    lines.push('No topology changes detected.');
    return lines.join('\n');
  }

  // Summary table
  lines.push('| File | Status | Buley | Regime |');
  lines.push('|:-----|:-------|:------|:-------|');

  for (const diff of diffs) {
    for (const func of diff.functions) {
      if (func.status === 'unchanged') continue;

      const buleyStr = func.delta
        ? `${func.baseBuley} -> ${func.headBuley} (${
            func.delta.buleyNumber >= 0 ? '+' : ''
          }${func.delta.buleyNumber})`
        : func.status === 'added'
        ? `${func.headBuley} (new)`
        : `${func.baseBuley} (removed)`;

      const regimeStr = func.regimeTransition
        ? `${func.regimeTransition.from} -> ${func.regimeTransition.to}`
        : func.headMetrics?.regime ?? '—';

      const statusEmoji =
        func.status === 'added'
          ? 'added'
          : func.status === 'removed'
          ? 'removed'
          : (func.delta?.buleyNumber ?? 0) > 0
          ? 'regressed'
          : 'improved';

      lines.push(
        `| \`${path.basename(
          func.filePath
        )}\` | ${statusEmoji} | ${buleyStr} | ${regimeStr} |`
      );
    }
  }

  if (totalRegressions.length > 0) {
    lines.push('');
    lines.push('### Regressions');
    for (const r of totalRegressions) {
      lines.push(`- ${r}`);
    }
  }

  if (totalImprovements.length > 0) {
    lines.push('');
    lines.push('### Improvements');
    for (const imp of totalImprovements) {
      lines.push(`- ${imp}`);
    }
  }

  lines.push('');
  lines.push('---');
  lines.push('*Generated by Gnosis Topological Linter*');

  return lines.join('\n');
}
