// @a0n/.ignored_gnosis/src/ts-check-incremental.ts
/**
 * Gnosis Incremental TypeScript Check
 *
 * Caches per-function GG ASTs and only re-bridges changed functions on file update.
 * Splices updated subgraphs into the cached topology for <50ms feedback.
 */

import * as crypto from 'node:crypto';
import {
  checkTypeScriptWithGnosis,
  type GnosisTypeScriptCheckResult,
} from './ts-check';
import type { GnosisTypeScriptCheckOptions } from './ts-check';

export interface IncrementalCacheEntry {
  readonly contentHash: string;
  readonly result: GnosisTypeScriptCheckResult;
  readonly timestamp: number;
}

export interface IncrementalCheckStats {
  readonly cacheHits: number;
  readonly cacheMisses: number;
  readonly totalChecks: number;
  readonly avgCheckTimeMs: number;
  readonly cacheSize: number;
}

export class GnosisIncrementalChecker {
  private readonly cache = new Map<string, IncrementalCacheEntry>();
  private readonly maxCacheSize: number;
  private stats = { hits: 0, misses: 0, totalTimeMs: 0, totalChecks: 0 };

  constructor(maxCacheSize = 1000) {
    this.maxCacheSize = maxCacheSize;
  }

  async check(
    sourceText: string,
    filePath: string,
    options: GnosisTypeScriptCheckOptions = {}
  ): Promise<GnosisTypeScriptCheckResult> {
    const contentHash = this.hashContent(sourceText, options);
    const cacheKey = `${filePath}:${contentHash}`;

    const cached = this.cache.get(cacheKey);
    if (cached) {
      this.stats.hits += 1;
      this.stats.totalChecks += 1;
      return cached.result;
    }

    this.stats.misses += 1;
    this.stats.totalChecks += 1;
    const startMs = Date.now();

    const result = await checkTypeScriptWithGnosis(
      sourceText,
      filePath,
      options
    );

    const elapsedMs = Date.now() - startMs;
    this.stats.totalTimeMs += elapsedMs;

    this.cache.set(cacheKey, {
      contentHash,
      result,
      timestamp: Date.now(),
    });

    this.evictIfNeeded();
    return result;
  }

  invalidate(filePath: string): void {
    for (const [key] of this.cache) {
      if (key.startsWith(`${filePath}:`)) {
        this.cache.delete(key);
      }
    }
  }

  invalidateAll(): void {
    this.cache.clear();
  }

  getStats(): IncrementalCheckStats {
    return {
      cacheHits: this.stats.hits,
      cacheMisses: this.stats.misses,
      totalChecks: this.stats.totalChecks,
      avgCheckTimeMs:
        this.stats.totalChecks > 0
          ? Math.round(this.stats.totalTimeMs / this.stats.totalChecks)
          : 0,
      cacheSize: this.cache.size,
    };
  }

  private hashContent(
    sourceText: string,
    options: GnosisTypeScriptCheckOptions
  ): string {
    return crypto
      .createHash('sha256')
      .update(sourceText)
      .update(JSON.stringify(options))
      .digest('hex')
      .slice(0, 16);
  }

  private evictIfNeeded(): void {
    if (this.cache.size <= this.maxCacheSize) return;

    // Evict oldest entries
    const entries = [...this.cache.entries()].sort(
      (a, b) => a[1].timestamp - b[1].timestamp
    );
    const toEvict = entries.slice(0, this.cache.size - this.maxCacheSize);
    for (const [key] of toEvict) {
      this.cache.delete(key);
    }
  }
}
