// @a0n/.ignored_gnosis/src/ts-check-watcher.ts
/**
 * Gnosis File Watcher
 *
 * Watches TypeScript files for changes and runs topology checks automatically.
 * Pushes results to registered listeners (SSE endpoints, IDE plugins, etc.)
 */

import * as fs from 'node:fs';
import * as path from 'node:path';
import { GnosisIncrementalChecker } from './ts-check-incremental';
import type {
  GnosisTypeScriptCheckResult,
  GnosisTypeScriptCheckOptions,
} from './ts-check';
import {
  generateAutofixSuggestions,
  type AutofixSuggestion,
} from './ts-check-autofix';

export interface WatcherEvent {
  readonly type: 'check-complete' | 'error' | 'file-change';
  readonly filePath: string;
  readonly timestamp: number;
  readonly result?: GnosisTypeScriptCheckResult;
  readonly autofixes?: readonly AutofixSuggestion[];
  readonly error?: string;
}

export type WatcherListener = (event: WatcherEvent) => void;

export interface WatcherOptions {
  readonly patterns?: readonly string[];
  readonly ignorePatterns?: readonly string[];
  readonly debounceMs?: number;
  readonly checkOptions?: GnosisTypeScriptCheckOptions;
  readonly enableAutofix?: boolean;
}

const DEFAULT_PATTERNS = ['**/*.ts', '**/*.tsx'];
const DEFAULT_IGNORE = [
  '**/node_modules/**',
  '**/dist/**',
  '**/.next/**',
  '**/build/**',
  '**/*.d.ts',
  '**/*.test.*',
  '**/*.spec.*',
  '**/__tests__/**',
];

export class GnosisFileWatcher {
  private readonly checker: GnosisIncrementalChecker;
  private readonly listeners = new Set<WatcherListener>();
  private readonly watchers = new Map<string, fs.FSWatcher>();
  private readonly debounceTimers = new Map<
    string,
    ReturnType<typeof setTimeout>
  >();
  private readonly options: Required<WatcherOptions>;
  private running = false;

  constructor(options: WatcherOptions = {}) {
    this.checker = new GnosisIncrementalChecker();
    this.options = {
      patterns: options.patterns ?? DEFAULT_PATTERNS,
      ignorePatterns: options.ignorePatterns ?? DEFAULT_IGNORE,
      debounceMs: options.debounceMs ?? 500,
      checkOptions: options.checkOptions ?? {},
      enableAutofix: options.enableAutofix ?? true,
    };
  }

  addListener(listener: WatcherListener): () => void {
    this.listeners.add(listener);
    return () => this.listeners.delete(listener);
  }

  private emit(event: WatcherEvent): void {
    for (const listener of this.listeners) {
      try {
        listener(event);
      } catch {
        // Listener errors don't break the watcher
      }
    }
  }

  async watchDirectory(dirPath: string): Promise<void> {
    if (this.running) return;
    this.running = true;

    const absDir = path.resolve(dirPath);

    // Skip initial full-workspace scan -- it type-checks every .ts file
    // in the monorepo which pegs CPU at 100%+ for minutes. Only check
    // files as they change going forward.

    // Watch for changes using fs.watch (recursive)
    try {
      const watcher = fs.watch(
        absDir,
        { recursive: true },
        (_eventType, filename) => {
          if (!filename) return;
          const fullPath = path.join(absDir, filename);

          if (!this.isTypeScriptFile(fullPath)) return;
          if (this.isIgnored(fullPath)) return;

          // Debounce
          const existing = this.debounceTimers.get(fullPath);
          if (existing) clearTimeout(existing);

          this.debounceTimers.set(
            fullPath,
            setTimeout(() => {
              this.debounceTimers.delete(fullPath);
              this.emit({
                type: 'file-change',
                filePath: fullPath,
                timestamp: Date.now(),
              });
              void this.checkFile(fullPath);
            }, this.options.debounceMs)
          );
        }
      );

      this.watchers.set(absDir, watcher);
    } catch {
      // Recursive watch not supported on all platforms; fall back to individual watches
    }
  }

  stop(): void {
    this.running = false;
    for (const [, watcher] of this.watchers) {
      watcher.close();
    }
    this.watchers.clear();
    for (const [, timer] of this.debounceTimers) {
      clearTimeout(timer);
    }
    this.debounceTimers.clear();
  }

  getStats() {
    return this.checker.getStats();
  }

  private async checkFile(filePath: string): Promise<void> {
    try {
      if (!fs.existsSync(filePath)) return;

      // polyglot:ignore RESOURCE_LEAK — readFileSync returns a string, no handle to release
      const sourceText = fs.readFileSync(filePath, 'utf-8');
      const result = await this.checker.check(
        sourceText,
        filePath,
        this.options.checkOptions
      );

      let autofixes: readonly AutofixSuggestion[] | undefined;
      if (this.options.enableAutofix) {
        autofixes = generateAutofixSuggestions(result, sourceText);
      }

      this.emit({
        type: 'check-complete',
        filePath,
        timestamp: Date.now(),
        result,
        ...(autofixes !== undefined ? { autofixes } : {}),
      });
    } catch (err) {
      this.emit({
        type: 'error',
        filePath,
        timestamp: Date.now(),
        error: err instanceof Error ? err.message : String(err),
      });
    }
  }

  public _findTypeScriptFiles(dirPath: string): string[] {
    const files: string[] = [];

    const walk = (dir: string): void => {
      let entries: fs.Dirent[];
      try {
        entries = fs.readdirSync(dir, { withFileTypes: true });
      } catch {
        return;
      }

      for (const entry of entries) {
        const fullPath = path.join(dir, entry.name);
        if (entry.isDirectory()) {
          if (!this.isIgnoredDir(entry.name)) {
            walk(fullPath);
          }
        } else if (
          entry.isFile() &&
          this.isTypeScriptFile(fullPath) &&
          !this.isIgnored(fullPath)
        ) {
          files.push(fullPath);
        }
      }
    };

    walk(dirPath);
    return files;
  }

  private isTypeScriptFile(filePath: string): boolean {
    return filePath.endsWith('.ts') || filePath.endsWith('.tsx');
  }

  private isIgnoredDir(name: string): boolean {
    return (
      name === 'node_modules' ||
      name === 'dist' ||
      name === '.next' ||
      name === 'build' ||
      name === '.bun' ||
      name === '.nx' ||
      name === 'coverage'
    );
  }

  private isIgnored(filePath: string): boolean {
    if (filePath.endsWith('.d.ts') || filePath.endsWith('.d.tsx')) return true;
    if (filePath.includes('.test.') || filePath.includes('.spec.')) return true;
    if (filePath.includes('__tests__')) return true;
    if (filePath.includes('node_modules')) return true;
    return false;
  }
}
