// @a0n/.ignored_gnosis/src/ts-check-cli.ts
#!/usr/bin/env bun
/**
 * Gnosis TypeScript Check CLI
 *
 * Usage:
 *   bun run open-source/gnosis/src/ts-check-cli.ts --file <path> [--json] [--sarif] [--max-buley N] [--target T] [--export NAME]
 */

import * as fs from 'node:fs';
import { checkTypeScriptWithGnosis } from './ts-check';
import type { RuntimeTarget } from './capabilities/index';

interface CliArgs {
  file: string | null;
  json: boolean;
  sarif: boolean;
  maxBuley: number | undefined;
  target: RuntimeTarget | undefined;
  exportName: string | undefined;
}

function parseArgs(argv: string[]): CliArgs {
  const result: CliArgs = {
    file: null,
    json: false,
    sarif: false,
    maxBuley: undefined,
    target: undefined,
    exportName: undefined,
  };

  for (let i = 0; i < argv.length; i += 1) {
    const arg = argv[i];
    if (arg === '--file' && i + 1 < argv.length) {
      result.file = argv[++i];
    } else if (arg === '--json') {
      result.json = true;
    } else if (arg === '--sarif') {
      result.sarif = true;
    } else if (arg === '--max-buley' && i + 1 < argv.length) {
      result.maxBuley = Number(argv[++i]);
    } else if (arg === '--target' && i + 1 < argv.length) {
      result.target = argv[++i] as RuntimeTarget;
    } else if (arg === '--export' && i + 1 < argv.length) {
      result.exportName = argv[++i];
    }
  }

  return result;
}

async function main(): Promise<void> {
  const args = parseArgs(process.argv.slice(2));

  if (!args.file) {
    console.error(
      'Usage: ts-check-cli --file <path> [--json] [--sarif] [--max-buley N] [--target T] [--export NAME]'
    );
    process.exit(1);
  }

  let sourceText: string;
  try {
    sourceText = fs.readFileSync(args.file, 'utf-8');
  } catch (err) {
    console.error(`Failed to read file: ${args.file}`);
    process.exit(1);
  }

  try {
    const result = await checkTypeScriptWithGnosis(sourceText, args.file, {
      maxBuley: args.maxBuley,
      target: args.target,
      exportName: args.exportName,
    });

    if (args.sarif) {
      console.log(JSON.stringify(result.sarif, null, 2));
      process.exit(result.ok ? 0 : 1);
    }

    if (args.json) {
      console.log(
        JSON.stringify(
          {
            ok: result.ok,
            diagnostics: result.diagnostics,
            metrics: result.metrics,
            topology: result.topology,
            ggSource: result.ggSource,
          },
          null,
          2
        )
      );
      process.exit(result.ok ? 0 : 1);
    }

    // Human-readable output
    console.log(`[gnosis ts-check] ${args.file}`);
    console.log(`  status: ${result.ok ? 'PASS' : 'FAIL'}`);
    console.log(`  buley-number: ${result.metrics.buleyNumber}`);
    console.log(`  wallace-number: ${result.metrics.wallaceNumber}`);
    console.log(`  regime: ${result.metrics.regime}`);
    console.log(`  beta1: ${result.metrics.beta1}`);
    console.log(`  quantum-index: ${result.metrics.quantumIndex}`);
    console.log(
      `  nodes=${result.metrics.nodeCount} edges=${result.metrics.edgeCount}`
    );

    if (result.diagnostics.length > 0) {
      console.log('  diagnostics:');
      for (const diagnostic of result.diagnostics) {
        console.log(
          `    [${diagnostic.level}] ${diagnostic.ruleId} ${diagnostic.message} (${diagnostic.line}:${diagnostic.column})`
        );
      }
    }

    process.exit(result.ok ? 0 : 1);
  } catch (err) {
    // Files that can't be bridged (no exported functions, unsupported patterns)
    // are not errors -- just no diagnostics
    const message = err instanceof Error ? err.message : String(err);
    if (args.json) {
      console.log(
        JSON.stringify(
          {
            ok: true,
            diagnostics: [],
            metrics: null,
            topology: null,
            ggSource: null,
            skipped: true,
            skipReason: message,
          },
          null,
          2
        )
      );
      process.exit(0);
    }
    if (args.sarif) {
      console.log(
        JSON.stringify(
          {
            version: '2.1.0',
            $schema: 'https://json.schemastore.org/sarif-2.1.0.json',
            runs: [
              {
                tool: {
                  driver: {
                    name: 'gnosis-ts-check',
                    informationUri: 'https://github.com/forkjoin-ai/gnosis',
                    rules: [],
                  },
                },
                results: [],
              },
            ],
          },
          null,
          2
        )
      );
      process.exit(0);
    }
    console.log(`[gnosis ts-check] ${args.file} -- skipped: ${message}`);
    process.exit(0);
  }
}

void main();
