// @a0n/.ignored_gnosis/src/ts-check-advanced.test.ts
import { describe, it, expect } from '@a0n/gnosis/test';
import { checkTypeScriptWithGnosis } from './ts-check';
import { GnosisIncrementalChecker } from './ts-check-incremental';
import {
  detectDeadBranches,
  generateAutofixSuggestions,
} from './ts-check-autofix';
import { computeTopologyDiff, formatPrComment } from './ts-check-diff';
import {
  generateLeanCertificate,
  verifyCertificate,
} from './ts-lean-certificate';
import {
  GnosisProbeCollector,
  createProbeWrapper,
  buildOtelAttributes,
} from './ts-runtime-probe';

const SIMPLE_TS = `
export async function orchestrate(input: unknown) {
  const result = await processData(input);
  return result;
}
`;

const PARALLEL_TS = `
export async function orchestrate(input: unknown) {
  const [a, b] = await Promise.all([fetchA(input), fetchB(input)]);
  const merged = await merge(a, b);
  return merged;
}
`;

const COMPLEX_TS = `
export async function complexHandler(input: unknown) {
  const a = await stepA(input);
  const b = await stepB(a);
  const c = await stepC(b);
  const d = await stepD(c);
  const e = await stepE(d);
  const f = await stepF(e);
  const g = await stepG(f);
  return g;
}
`;

const CHANGED_TS = `
export async function orchestrate(input: unknown) {
  const a = await stepA(input);
  const b = await stepB(a);
  const c = await stepC(b);
  return c;
}
`;

describe('GnosisIncrementalChecker', () => {
  it('caches results by content hash', async () => {
    const checker = new GnosisIncrementalChecker();

    const r1 = await checker.check(SIMPLE_TS, 'test.ts');
    const r2 = await checker.check(SIMPLE_TS, 'test.ts');

    expect(r1.metrics.buleyNumber).toBe(r2.metrics.buleyNumber);
    const stats = checker.getStats();
    expect(stats.cacheHits).toBe(1);
    expect(stats.cacheMisses).toBe(1);
  });

  it('invalidates cache for specific file', async () => {
    const checker = new GnosisIncrementalChecker();

    await checker.check(SIMPLE_TS, 'test.ts');
    checker.invalidate('test.ts');
    await checker.check(SIMPLE_TS, 'test.ts');

    const stats = checker.getStats();
    expect(stats.cacheMisses).toBe(2);
  });

  it('tracks cache size', async () => {
    const checker = new GnosisIncrementalChecker();

    await checker.check(SIMPLE_TS, 'a.ts');
    await checker.check(PARALLEL_TS, 'b.ts');

    const stats = checker.getStats();
    expect(stats.cacheSize).toBe(2);
  });
});

describe('detectDeadBranches', () => {
  it('returns empty for a linear topology', async () => {
    const result = await checkTypeScriptWithGnosis(SIMPLE_TS, 'test.ts');
    const dead = detectDeadBranches(result);
    expect(dead).toEqual([]);
  });

  it('returns empty for parallel topology', async () => {
    const result = await checkTypeScriptWithGnosis(PARALLEL_TS, 'test.ts');
    const dead = detectDeadBranches(result);
    expect(dead).toEqual([]);
  });
});

describe('generateAutofixSuggestions', () => {
  it('returns suggestions for a simple function', async () => {
    const result = await checkTypeScriptWithGnosis(SIMPLE_TS, 'test.ts');
    const fixes = generateAutofixSuggestions(result, SIMPLE_TS);
    expect(Array.isArray(fixes)).toBe(true);
  });

  it('suggests parallelization for sequential chains', async () => {
    const result = await checkTypeScriptWithGnosis(COMPLEX_TS, 'test.ts');
    const fixes = generateAutofixSuggestions(result, COMPLEX_TS);

    const parallelSuggestions = fixes.filter(
      (f) => f.kind === 'extract-parallel'
    );
    // May or may not find independent operations depending on analysis
    expect(Array.isArray(parallelSuggestions)).toBe(true);
  });

  it('includes severity on all suggestions', async () => {
    const result = await checkTypeScriptWithGnosis(COMPLEX_TS, 'test.ts');
    const fixes = generateAutofixSuggestions(result, COMPLEX_TS);

    for (const fix of fixes) {
      expect(['info', 'suggestion', 'refactor']).toContain(fix.severity);
      expect(fix.ruleId).toBeTruthy();
      expect(fix.message).toBeTruthy();
      expect(fix.line).toBeGreaterThan(0);
    }
  });
});

describe('computeTopologyDiff', () => {
  it('detects no change for identical sources', async () => {
    const diff = await computeTopologyDiff(SIMPLE_TS, SIMPLE_TS, 'test.ts');
    expect(diff.functions[0]?.status).toBe('unchanged');
    expect(diff.regressions).toEqual([]);
    expect(diff.improvements).toEqual([]);
  });

  it('detects changes between different sources', async () => {
    const diff = await computeTopologyDiff(SIMPLE_TS, CHANGED_TS, 'test.ts');
    expect(diff.functions[0]?.status).toBe('changed');
    expect(diff.functions[0]?.delta).toBeDefined();
    expect(typeof diff.functions[0]?.delta?.buleyNumber).toBe('number');
  });

  it('detects new functions', async () => {
    const diff = await computeTopologyDiff('', SIMPLE_TS, 'test.ts');
    expect(diff.functions[0]?.status).toBe('added');
  });

  it('detects removed functions', async () => {
    const diff = await computeTopologyDiff(SIMPLE_TS, '', 'test.ts');
    expect(diff.functions[0]?.status).toBe('removed');
  });

  it('reports regime transitions', async () => {
    const diff = await computeTopologyDiff(SIMPLE_TS, CHANGED_TS, 'test.ts');
    if (diff.functions[0]?.regimeTransition) {
      expect(typeof diff.functions[0].regimeTransition.from).toBe('string');
      expect(typeof diff.functions[0].regimeTransition.to).toBe('string');
      expect(typeof diff.functions[0].regimeTransition.degraded).toBe(
        'boolean'
      );
    }
  });
});

describe('formatPrComment', () => {
  it('formats a PR comment with diff data', async () => {
    const diff = await computeTopologyDiff(SIMPLE_TS, CHANGED_TS, 'test.ts');
    const comment = formatPrComment([diff]);
    expect(comment).toContain('Gnosis Topology Diff');
    expect(typeof comment).toBe('string');
    expect(comment.length).toBeGreaterThan(0);
  });

  it('handles empty diffs', () => {
    const comment = formatPrComment([]);
    expect(comment).toContain('No topology changes');
  });
});

describe('generateLeanCertificate', () => {
  it('generates a certificate for a passing function', async () => {
    const cert = await generateLeanCertificate(SIMPLE_TS, 'test.ts');
    expect(cert).not.toBeNull();
    expect(cert!.verified).toBe(true);
    expect(cert!.leanSource).toContain('namespace Gnosis.Certificate');
    expect(cert!.leanSource).toContain('theorem beta1_bounded');
    expect(cert!.leanSource).toContain('theorem finite_state_space');
    expect(cert!.buleyNumber).toBeGreaterThan(0);
  });

  it('includes correct metrics in the certificate', async () => {
    const cert = await generateLeanCertificate(SIMPLE_TS, 'test.ts');
    expect(cert).not.toBeNull();
    expect(typeof cert!.buleyNumber).toBe('number');
    expect(typeof cert!.wallaceNumber).toBe('number');
    expect(typeof cert!.regime).toBe('string');
    expect(typeof cert!.nodeCount).toBe('number');
    expect(typeof cert!.edgeCount).toBe('number');
  });
});

describe('verifyCertificate', () => {
  it('verifies a valid certificate', async () => {
    const cert = await generateLeanCertificate(SIMPLE_TS, 'test.ts');
    expect(cert).not.toBeNull();

    const verification = await verifyCertificate(SIMPLE_TS, 'test.ts', cert!);
    expect(verification.valid).toBe(true);
  });

  it('rejects a certificate when source changes', async () => {
    const cert = await generateLeanCertificate(SIMPLE_TS, 'test.ts');
    expect(cert).not.toBeNull();

    const verification = await verifyCertificate(CHANGED_TS, 'test.ts', cert!);
    expect(verification.valid).toBe(false);
    expect(verification.reason.length).toBeGreaterThan(0);
  });
});

describe('GnosisProbeCollector', () => {
  it('records probe sessions', () => {
    const collector = new GnosisProbeCollector();
    const sessionId = collector.startSession('test', 'test.ts');

    collector.recordEvent(sessionId, { nodeId: 'entry', kind: 'enter' });
    collector.recordEvent(sessionId, { nodeId: 'call_1', kind: 'enter' });
    collector.recordEvent(sessionId, { nodeId: 'call_1', kind: 'exit' });
    collector.recordEvent(sessionId, { nodeId: 'entry', kind: 'exit' });
    collector.endSession(sessionId);

    expect(collector.getActiveSessions()).toBe(0);
    expect(collector.getCompletedCount()).toBe(1);
  });

  it('computes runtime metrics', () => {
    const collector = new GnosisProbeCollector();

    for (let i = 0; i < 5; i++) {
      const sessionId = collector.startSession('test', 'test.ts');
      collector.recordEvent(sessionId, { nodeId: 'entry', kind: 'enter' });
      collector.recordEvent(sessionId, { nodeId: 'entry', kind: 'exit' });
      collector.endSession(sessionId);
    }

    const metrics = collector.getMetrics('test', 'test.ts');
    expect(metrics.invocationCount).toBe(5);
    expect(metrics.functionName).toBe('test');
    expect(metrics.filePath).toBe('test.ts');
    expect(['laminar', 'transitional', 'turbulent']).toContain(
      metrics.empiricalRegime
    );
  });

  it('computes wallace divergence', async () => {
    const collector = new GnosisProbeCollector();

    const sessionId = collector.startSession('test', 'test.ts');
    collector.recordEvent(sessionId, { nodeId: 'entry', kind: 'enter' });
    collector.recordEvent(sessionId, { nodeId: 'entry', kind: 'exit' });
    collector.endSession(sessionId);

    const divergence = collector.computeWallaceDivergence(
      'test',
      'test.ts',
      0.5
    );
    expect(typeof divergence.divergence).toBe('number');
    expect(['pessimistic-model', 'aligned', 'latent-risk']).toContain(
      divergence.interpretation
    );
  });

  it('tracks branch frequencies', () => {
    const collector = new GnosisProbeCollector();

    for (let i = 0; i < 10; i++) {
      const sessionId = collector.startSession('test', 'test.ts');
      collector.recordEvent(sessionId, {
        nodeId: i < 7 ? 'branch_a' : 'branch_b',
        kind: 'fork-branch',
        metadata: { forkNodeId: 'fork_1' },
      });
      collector.endSession(sessionId);
    }

    const metrics = collector.getMetrics('test', 'test.ts');
    expect(metrics.branchFrequencies.length).toBeGreaterThan(0);

    const branchA = metrics.branchFrequencies.find(
      (f) => f.branchNodeId === 'branch_a'
    );
    expect(branchA).toBeDefined();
    expect(branchA!.selectionCount).toBe(7);
  });
});

describe('createProbeWrapper', () => {
  it('wraps a function with probes', async () => {
    const collector = new GnosisProbeCollector();
    const wrapper = createProbeWrapper(collector, 'test', 'test.ts', [
      'entry',
      'call_1',
    ]);

    const result = await wrapper(async () => 42);
    expect(result).toBe(42);
    expect(collector.getCompletedCount()).toBe(1);
  });

  it('records errors in probes', async () => {
    const collector = new GnosisProbeCollector();
    const wrapper = createProbeWrapper(collector, 'test', 'test.ts', ['entry']);

    await expect(
      wrapper(async () => {
        throw new Error('test error');
      })
    ).rejects.toThrow('test error');

    expect(collector.getCompletedCount()).toBe(1);
  });
});

describe('buildOtelAttributes', () => {
  it('builds correct OTel span attributes', () => {
    const attrs = buildOtelAttributes(
      'node_1',
      'call',
      'orchestrate',
      'handler.ts',
      12.5,
      0.3,
      'laminar'
    );

    expect(attrs['gnosis.node.id']).toBe('node_1');
    expect(attrs['gnosis.node.kind']).toBe('call');
    expect(attrs['gnosis.function']).toBe('orchestrate');
    expect(attrs['gnosis.file']).toBe('handler.ts');
    expect(attrs['gnosis.buley']).toBe(12.5);
    expect(attrs['gnosis.wallace']).toBe(0.3);
    expect(attrs['gnosis.regime']).toBe('laminar');
  });
});
