// @a0n/.ignored_gnosis/src/ts-check.ts
/**
 * Gnosis TypeScript Check Engine
 *
 * Unified pipeline: TS source -> GG topology -> formal check -> flat diagnostics.
 * Bridges the ts-bridge, analysis, and SARIF modules into a single check API.
 */

import {
  compileTypeScriptToGnosis,
  type GnosisTypeScriptBridgeNodePlan,
  type GnosisTypeScriptBridgeResult,
  type GnosisTypeScriptBridgeSourceLocation,
} from './ts-bridge';
import {
  analyzeGnosisSource,
  formatGnosisViolations,
  type GnosisComplexityReport,
  type GnosisSteeringRegime,
} from './analysis';
import {
  analyzePolyglotSourceString,
  type PolyglotDiagnosticFinding,
  type PolyglotFunctionAnalysis,
  type PolyglotGhostMassLedger,
} from './polyglot-bridge';
import { ggReportToSarif } from './sarif';
import type { RuntimeTarget } from './capabilities/index';

export interface GnosisTypeScriptCheckOptions {
  readonly exportName?: string;
  readonly maxBuley?: number;
  readonly target?: RuntimeTarget;
}

export type GnosisTypeScriptCheckDiagnosticLevel = 'error' | 'warning' | 'note';

export interface GnosisTypeScriptCheckDiagnostic {
  readonly ruleId: string;
  readonly level: GnosisTypeScriptCheckDiagnosticLevel;
  readonly message: string;
  readonly line: number;
  readonly column: number;
  readonly endLine?: number;
  readonly endColumn?: number;
  readonly theoremAnchor?: string;
  readonly ghostMass?: number;
}

export interface GnosisTypeScriptCheckTopologyNode {
  readonly id: string;
  readonly label: string;
  readonly kind: string;
  readonly x: number;
  readonly y: number;
  readonly sourceLocation?: GnosisTypeScriptBridgeSourceLocation;
}

export interface GnosisTypeScriptCheckTopologyEdge {
  readonly from: string;
  readonly to: string;
  readonly type: string;
}

export interface GnosisTypeScriptCheckTopologyGraph {
  readonly nodes: readonly GnosisTypeScriptCheckTopologyNode[];
  readonly edges: readonly GnosisTypeScriptCheckTopologyEdge[];
}

export interface GnosisTypeScriptCheckMetrics {
  readonly buleyNumber: number;
  readonly wallaceNumber: number;
  readonly regime: GnosisSteeringRegime;
  readonly beta1: number;
  readonly quantumIndex: number;
  readonly nodeCount: number;
  readonly edgeCount: number;
  readonly ghostMass: number;
  readonly leakDebt: number;
}

export interface GnosisGhostMassLedger {
  readonly forkedMass: number;
  readonly liveMass: number;
  readonly foldedMass: number;
  readonly ventedMass: number;
  readonly accountedMass: number;
  readonly ghostMass: number;
  readonly leakDebt: number;
}

type MutableGhostMassLedger = {
  -readonly [Key in keyof GnosisGhostMassLedger]: GnosisGhostMassLedger[Key];
};

export interface GnosisTypeScriptCheckResult {
  readonly ok: boolean;
  readonly diagnostics: readonly GnosisTypeScriptCheckDiagnostic[];
  readonly metrics: GnosisTypeScriptCheckMetrics;
  readonly topology: GnosisTypeScriptCheckTopologyGraph;
  readonly sarif: unknown;
  readonly ggSource: string;
  readonly ghostMassLedger: GnosisGhostMassLedger;
}

function buildNodeLocationMap(
  nodePlans: readonly GnosisTypeScriptBridgeNodePlan[]
): Map<string, GnosisTypeScriptBridgeSourceLocation> {
  const map = new Map<string, GnosisTypeScriptBridgeSourceLocation>();
  for (const plan of nodePlans) {
    if (plan.sourceLocation) {
      map.set(plan.nodeId, plan.sourceLocation);
    }
  }
  return map;
}

function defaultLocation(): { line: number; column: number } {
  return { line: 1, column: 1 };
}

function emptyGhostMassLedger(): GnosisGhostMassLedger {
  return {
    forkedMass: 0,
    liveMass: 0,
    foldedMass: 0,
    ventedMass: 0,
    accountedMass: 0,
    ghostMass: 0,
    leakDebt: 0,
  };
}

function normalizePolyglotCode(code: string): string {
  if (code.includes('_')) {
    return code.toUpperCase();
  }

  return code
    .replace(/([a-z0-9])([A-Z])/g, '$1_$2')
    .replace(/-/g, '_')
    .toUpperCase();
}

function mapPolyglotSeverity(
  severity: PolyglotDiagnosticFinding['severity']
): GnosisTypeScriptCheckDiagnosticLevel {
  switch (severity) {
    case 'Error':
      return 'error';
    case 'Warning':
      return 'warning';
    default:
      return 'note';
  }
}

function toGhostMassLedger(
  ledger?: PolyglotGhostMassLedger
): GnosisGhostMassLedger {
  if (!ledger) {
    return emptyGhostMassLedger();
  }

  return {
    forkedMass: ledger.forked_mass,
    liveMass: ledger.live_mass,
    foldedMass: ledger.folded_mass,
    ventedMass: ledger.vented_mass,
    accountedMass: ledger.accounted_mass,
    ghostMass: ledger.ghost_mass,
    leakDebt: ledger.leak_debt,
  };
}

function mergeGhostMassLedgers(
  functions: readonly PolyglotFunctionAnalysis[]
): GnosisGhostMassLedger {
  const aggregate: MutableGhostMassLedger = { ...emptyGhostMassLedger() };

  for (const func of functions) {
    const ledger = toGhostMassLedger(func.ghostMassLedger);
    aggregate.forkedMass += ledger.forkedMass;
    aggregate.liveMass += ledger.liveMass;
    aggregate.foldedMass += ledger.foldedMass;
    aggregate.ventedMass += ledger.ventedMass;
    aggregate.accountedMass += ledger.accountedMass;
    aggregate.ghostMass += ledger.ghostMass;
    aggregate.leakDebt += ledger.leakDebt;
  }

  return aggregate;
}

function polyglotDiagnosticLocation(
  func: PolyglotFunctionAnalysis,
  diagnostic: PolyglotDiagnosticFinding
): GnosisTypeScriptBridgeSourceLocation {
  if (diagnostic.location) {
    return {
      line: diagnostic.location.start_line,
      column: diagnostic.location.start_column,
      endLine: diagnostic.location.end_line,
      endColumn: diagnostic.location.end_column,
    };
  }

  for (const nodeId of diagnostic.related_nodes) {
    const span = func.sourceMap.get(nodeId);
    if (span) {
      return {
        line: span.start_line,
        column: span.start_column,
        endLine: span.end_line,
        endColumn: span.end_column,
      };
    }
  }

  return defaultLocation();
}

function buildPolyglotDiagnostics(
  functions: readonly PolyglotFunctionAnalysis[]
): GnosisTypeScriptCheckDiagnostic[] {
  const diagnostics: GnosisTypeScriptCheckDiagnostic[] = [];
  const seen = new Set<string>();

  for (const func of functions) {
    for (const diagnostic of func.diagnostics) {
      const loc = polyglotDiagnosticLocation(func, diagnostic);
      const normalizedCode = normalizePolyglotCode(diagnostic.code);
      const key = `${normalizedCode}:${loc.line}:${loc.column}:${diagnostic.message}`;
      if (seen.has(key)) {
        continue;
      }
      seen.add(key);

      diagnostics.push({
        ruleId: `gnosis.polyglot.${normalizedCode}`,
        level: mapPolyglotSeverity(diagnostic.severity),
        message: diagnostic.message,
        line: loc.line,
        column: loc.column,
        ...(loc.endLine !== undefined ? { endLine: loc.endLine } : {}),
        ...(loc.endColumn !== undefined ? { endColumn: loc.endColumn } : {}),
        ...(diagnostic.theorem_anchor != null
          ? { theoremAnchor: diagnostic.theorem_anchor }
          : {}),
        ...(diagnostic.ghost_mass != null
          ? { ghostMass: diagnostic.ghost_mass }
          : {}),
      });
    }
  }

  return diagnostics;
}

function buildCounterfeitLivenessDiagnostic(
  ledger: GnosisGhostMassLedger,
  diagnostics: readonly GnosisTypeScriptCheckDiagnostic[]
): GnosisTypeScriptCheckDiagnostic | null {
  if (ledger.ghostMass === 0) {
    return null;
  }

  const firstLeakLocation =
    diagnostics.find((diagnostic) =>
      diagnostic.ruleId.startsWith('gnosis.polyglot.')
    ) ?? null;
  const loc: {
    line: number;
    column: number;
    endLine?: number;
    endColumn?: number;
  } = firstLeakLocation
    ? {
        line: firstLeakLocation.line,
        column: firstLeakLocation.column,
        ...(firstLeakLocation.endLine !== undefined
          ? { endLine: firstLeakLocation.endLine }
          : {}),
        ...(firstLeakLocation.endColumn !== undefined
          ? { endColumn: firstLeakLocation.endColumn }
          : {}),
      }
    : defaultLocation();

  return {
    ruleId: 'gnosis/counterfeit-liveness',
    level: 'error',
    message:
      `Counterfeit liveness detected: ghost mass ${ledger.ghostMass} remains after fold accounting ` +
      `(forked=${ledger.forkedMass}, folded=${ledger.foldedMass}, leak_witnesses=${ledger.ventedMass}).`,
    line: loc.line,
    column: loc.column,
    ...(loc.endLine !== undefined ? { endLine: loc.endLine } : {}),
    ...(loc.endColumn !== undefined ? { endColumn: loc.endColumn } : {}),
    theoremAnchor: 'memory_leak_master',
    ghostMass: ledger.ghostMass,
  };
}

function appendDiagnosticsToSarif(
  sarif: ReturnType<typeof ggReportToSarif>,
  filePath: string,
  diagnostics: readonly GnosisTypeScriptCheckDiagnostic[]
): ReturnType<typeof ggReportToSarif> {
  const log = JSON.parse(JSON.stringify(sarif)) as ReturnType<
    typeof ggReportToSarif
  > & {
    runs?: Array<{
      tool?: {
        driver?: {
          rules?: Array<{ id: string; shortDescription: { text: string } }>;
        };
      };
      results?: Array<Record<string, unknown>>;
    }>;
  };
  const run = log.runs?.[0];
  if (!run?.tool?.driver) {
    return sarif;
  }

  const rules = (run.tool.driver.rules ??= []);
  const results = (run.results ??= []);
  const existingRules = new Set(rules.map((rule) => rule.id));

  for (const diagnostic of diagnostics) {
    if (!existingRules.has(diagnostic.ruleId)) {
      rules.push({
        id: diagnostic.ruleId,
        shortDescription: { text: diagnostic.ruleId },
      });
      existingRules.add(diagnostic.ruleId);
    }

    const result: Record<string, unknown> = {
      ruleId: diagnostic.ruleId,
      level: diagnostic.level,
      message: { text: diagnostic.message },
      locations: [
        {
          physicalLocation: {
            artifactLocation: { uri: filePath },
            region: {
              startLine: diagnostic.line,
              startColumn: diagnostic.column,
              endLine: diagnostic.endLine ?? diagnostic.line,
              endColumn: diagnostic.endColumn ?? diagnostic.column,
            },
          },
        },
      ],
    };

    if (diagnostic.theoremAnchor || diagnostic.ghostMass !== undefined) {
      result['properties'] = {
        ...(diagnostic.theoremAnchor
          ? { theoremAnchor: diagnostic.theoremAnchor }
          : {}),
        ...(diagnostic.ghostMass !== undefined
          ? { ghostMass: diagnostic.ghostMass }
          : {}),
      };
    }

    results.push(result);
  }

  return log;
}

function buildTopologyGraph(
  bridgeResult: GnosisTypeScriptBridgeResult,
  _report: GnosisComplexityReport
): GnosisTypeScriptCheckTopologyGraph {
  const locationMap = buildNodeLocationMap(bridgeResult.nodePlans);
  const plansByNodeId = new Map<string, GnosisTypeScriptBridgeNodePlan>();
  for (const plan of bridgeResult.nodePlans) {
    plansByNodeId.set(plan.nodeId, plan);
  }

  const layerAssignment = new Map<string, number>();
  let layerIndex = 0;
  for (const wave of bridgeResult.schedule) {
    for (const nodeId of wave.nodeIds) {
      layerAssignment.set(nodeId, layerIndex);
    }
    layerIndex += 1;
  }

  const nodesPerLayer = new Map<number, number>();
  for (const [, layer] of layerAssignment) {
    nodesPerLayer.set(layer, (nodesPerLayer.get(layer) ?? 0) + 1);
  }
  const layerCounters = new Map<number, number>();
  const renderPlanLabel = (plan: GnosisTypeScriptBridgeNodePlan): string => {
    if (plan.kind !== 'call') {
      return plan.kind;
    }

    if (plan.callee.kind === 'runtime' || plan.callee.kind === 'binding') {
      return plan.callee.name;
    }

    if (plan.callee.kind === 'member') {
      return plan.callee.property;
    }

    if (plan.callee.kind === 'index') {
      return String(plan.callee.index);
    }

    return plan.kind;
  };

  const nodes: GnosisTypeScriptCheckTopologyNode[] = [];
  for (const plan of bridgeResult.nodePlans) {
    const layer = layerAssignment.get(plan.nodeId) ?? 0;
    const indexInLayer = layerCounters.get(layer) ?? 0;
    layerCounters.set(layer, indexInLayer + 1);
    const totalInLayer = nodesPerLayer.get(layer) ?? 1;
    const xSpread =
      totalInLayer > 1 ? (indexInLayer / (totalInLayer - 1)) * 10 - 5 : 0;
    const sourceLocation = locationMap.get(plan.nodeId);

    nodes.push({
      id: plan.nodeId,
      label: renderPlanLabel(plan),
      kind: plan.kind,
      x: xSpread,
      y: layer * -3,
      ...(sourceLocation !== undefined ? { sourceLocation } : {}),
    });
  }

  const edges: GnosisTypeScriptCheckTopologyEdge[] = [];
  for (const edge of bridgeResult.ast.edges) {
    for (const sourceId of edge.sourceIds) {
      for (const targetId of edge.targetIds) {
        edges.push({
          from: sourceId,
          to: targetId,
          type: edge.type,
        });
      }
    }
  }

  return { nodes, edges };
}

/**
 * Handles the gnosis check Type Script With Gnosis workflow.
 */
export async function checkTypeScriptWithGnosis(
  sourceText: string,
  filePath: string,
  options: GnosisTypeScriptCheckOptions = {}
): Promise<GnosisTypeScriptCheckResult> {
  const bridgeResult = compileTypeScriptToGnosis(sourceText, {
    sourceFilePath: filePath,
    ...(options.exportName !== undefined
      ? { exportName: options.exportName }
      : {}),
  });

  const report = await analyzeGnosisSource(bridgeResult.ggSource, {
    ...(options.target !== undefined ? { target: options.target } : {}),
  });
  const polyglotAnalysis = await analyzePolyglotSourceString(
    sourceText,
    filePath
  ).catch(() => null);

  const formattedViolations = formatGnosisViolations(report.correctness);
  let sarif = ggReportToSarif(
    filePath,
    report,
    formattedViolations,
    options.maxBuley ?? null
  );

  const locationMap = buildNodeLocationMap(bridgeResult.nodePlans);
  const diagnostics: GnosisTypeScriptCheckDiagnostic[] = [];

  for (const violation of formattedViolations) {
    const loc = locationMap.values().next().value ?? defaultLocation();
    diagnostics.push({
      ruleId: 'gnosis/formal-violation',
      level: 'error',
      message: violation,
      line: loc.line,
      column: loc.column,
    });
  }

  if (options.maxBuley !== undefined && report.buleyNumber > options.maxBuley) {
    diagnostics.push({
      ruleId: 'gnosis/buley-threshold',
      level: 'warning',
      message: `Buley number ${report.buleyNumber} exceeds threshold ${options.maxBuley}`,
      line: 1,
      column: 1,
    });
  }

  for (const issue of report.capabilities.issues) {
    const loc = issue.nodeId
      ? locationMap.get(issue.nodeId) ?? defaultLocation()
      : defaultLocation();
    diagnostics.push({
      ruleId: `gnosis/capability-${issue.capability}`,
      level: issue.severity === 'error' ? 'error' : 'warning',
      message: `${issue.message} (target=${issue.target})`,
      line: loc.line,
      column: loc.column,
    });
  }

  const regime = report.steering.regime;
  if (regime === 'turbulent') {
    diagnostics.push({
      ruleId: 'gnosis/steering-turbulent',
      level: 'warning',
      message: `Steering regime is turbulent (wallace=${report.steering.wallaceNumber}, bias=${report.steering.leanInBias})`,
      line: 1,
      column: 1,
    });
  }

  const ghostMassLedger = polyglotAnalysis
    ? mergeGhostMassLedgers(polyglotAnalysis.functions)
    : emptyGhostMassLedger();
  const polyglotDiagnostics = polyglotAnalysis
    ? buildPolyglotDiagnostics(polyglotAnalysis.functions)
    : [];
  diagnostics.push(...polyglotDiagnostics);

  const counterfeitLiveness = buildCounterfeitLivenessDiagnostic(
    ghostMassLedger,
    polyglotDiagnostics
  );
  if (counterfeitLiveness) {
    diagnostics.push(counterfeitLiveness);
  }

  sarif = appendDiagnosticsToSarif(
    sarif,
    filePath,
    diagnostics.filter(
      (diagnostic) =>
        diagnostic.ruleId.startsWith('gnosis.polyglot.') ||
        diagnostic.ruleId === 'gnosis/counterfeit-liveness'
    )
  );

  const topology = buildTopologyGraph(bridgeResult, report);

  const metrics: GnosisTypeScriptCheckMetrics = {
    buleyNumber: report.buleyNumber,
    wallaceNumber: report.steering.wallaceNumber,
    regime,
    beta1: report.topology.structuralBeta1,
    quantumIndex: report.quantum.quantumIndex,
    nodeCount: report.topology.nodeCount,
    edgeCount: report.topology.edgeCount,
    ghostMass: ghostMassLedger.ghostMass,
    leakDebt: ghostMassLedger.leakDebt,
  };

  return {
    ok: diagnostics.filter((d) => d.level === 'error').length === 0,
    diagnostics,
    metrics,
    topology,
    sarif,
    ggSource: bridgeResult.ggSource,
    ghostMassLedger,
  };
}
