// @a0n/.ignored_gnosis/src/ts-to-gg.ts
/**
 * TS → GG Transpiler -- heuristic TypeScript to GG topology extraction
 *
 * Detects fork/race/fold patterns in TypeScript source and emits
 * GG topology text. This is the 80% case -- not a full compiler,
 * but enough to capture the concurrency structure of most code.
 *
 * Pattern mappings:
 *   Promise.all([...])    → FORK
 *   Promise.race([...])   → RACE
 *   Promise.allSettled([]) → FORK
 *   await expr            → FOLD
 *   try { } catch { }     → FOLD + VENT
 *   if/else               → FORK + RACE
 *   for/while             → traced feedback loop
 *   throw                 → VENT
 *
 * The Betty compiler then verifies the topology is well-formed.
 */

interface ExtractedNode {
  id: string;
  label: string;
  line: number;
}

interface ExtractedEdge {
  source: string;
  target: string;
  type: 'FORK' | 'RACE' | 'FOLD' | 'VENT' | 'PROCESS';
  line: number;
}

export interface ExtractionResult {
  ggSource: string;
  nodes: ExtractedNode[];
  edges: ExtractedEdge[];
  stats: {
    forks: number;
    races: number;
    folds: number;
    vents: number;
    processes: number;
    linesScanned: number;
  };
}

/**
 * Extract GG topology from TypeScript source.
 */
export function extractTopology(
  tsSource: string,
  filename = 'input.ts'
): ExtractionResult {
  const lines = tsSource.split('\n');
  const nodes: ExtractedNode[] = [];
  const edges: ExtractedEdge[] = [];
  let nodeCounter = 0;

  function addNode(label: string, line: number): string {
    const id = `n${nodeCounter++}`;
    nodes.push({ id, label, line });
    return id;
  }

  // Entry node
  const entryId = addNode('entry', 0);
  let prevId = entryId;

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i].trim();
    const lineNum = i + 1;

    // Promise.all → FORK
    if (/Promise\.all\s*\(/.test(line)) {
      const forkId = addNode(`promise_all_L${lineNum}`, lineNum);
      edges.push({
        source: prevId,
        target: forkId,
        type: 'FORK',
        line: lineNum,
      });

      // Extract array items as parallel branches
      const match = line.match(/Promise\.all\s*\(\s*\[/);
      if (match) {
        const branchA = addNode(`branch_a_L${lineNum}`, lineNum);
        const branchB = addNode(`branch_b_L${lineNum}`, lineNum);
        edges.push({
          source: forkId,
          target: branchA,
          type: 'PROCESS',
          line: lineNum,
        });
        edges.push({
          source: forkId,
          target: branchB,
          type: 'PROCESS',
          line: lineNum,
        });

        const joinId = addNode(`join_L${lineNum}`, lineNum);
        edges.push({
          source: branchA,
          target: joinId,
          type: 'FOLD',
          line: lineNum,
        });
        edges.push({
          source: branchB,
          target: joinId,
          type: 'FOLD',
          line: lineNum,
        });
        prevId = joinId;
      } else {
        prevId = forkId;
      }
      continue;
    }

    // Promise.race → RACE
    if (/Promise\.race\s*\(/.test(line)) {
      const raceId = addNode(`promise_race_L${lineNum}`, lineNum);
      edges.push({
        source: prevId,
        target: raceId,
        type: 'FORK',
        line: lineNum,
      });

      const pathA = addNode(`racer_a_L${lineNum}`, lineNum);
      const pathB = addNode(`racer_b_L${lineNum}`, lineNum);
      edges.push({
        source: raceId,
        target: pathA,
        type: 'PROCESS',
        line: lineNum,
      });
      edges.push({
        source: raceId,
        target: pathB,
        type: 'PROCESS',
        line: lineNum,
      });

      const winnerId = addNode(`winner_L${lineNum}`, lineNum);
      edges.push({
        source: pathA,
        target: winnerId,
        type: 'RACE',
        line: lineNum,
      });
      edges.push({
        source: pathB,
        target: winnerId,
        type: 'RACE',
        line: lineNum,
      });
      prevId = winnerId;
      continue;
    }

    // await → FOLD (irreversible commitment to result)
    if (
      /\bawait\b/.test(line) &&
      !/Promise\.(all|race|allSettled)/.test(line)
    ) {
      const awaitId = addNode(`await_L${lineNum}`, lineNum);
      edges.push({
        source: prevId,
        target: awaitId,
        type: 'FOLD',
        line: lineNum,
      });
      prevId = awaitId;
      continue;
    }

    // try { → start of fold+vent pair
    if (/\btry\s*\{/.test(line)) {
      const tryId = addNode(`try_L${lineNum}`, lineNum);
      edges.push({
        source: prevId,
        target: tryId,
        type: 'FORK',
        line: lineNum,
      });
      prevId = tryId;
      continue;
    }

    // catch { → VENT (error path)
    if (/\bcatch\s*[({]/.test(line)) {
      const catchId = addNode(`catch_L${lineNum}`, lineNum);
      edges.push({
        source: prevId,
        target: catchId,
        type: 'VENT',
        line: lineNum,
      });
      prevId = catchId;
      continue;
    }

    // throw → VENT
    if (/\bthrow\b/.test(line)) {
      const throwId = addNode(`throw_L${lineNum}`, lineNum);
      edges.push({
        source: prevId,
        target: throwId,
        type: 'VENT',
        line: lineNum,
      });
      // Don't update prevId -- throw terminates this path
      continue;
    }

    // if ( → FORK
    if (/\bif\s*\(/.test(line)) {
      const ifId = addNode(`if_L${lineNum}`, lineNum);
      edges.push({ source: prevId, target: ifId, type: 'FORK', line: lineNum });
      prevId = ifId;
      continue;
    }

    // else → second branch of FORK
    if (/\belse\b/.test(line)) {
      const elseId = addNode(`else_L${lineNum}`, lineNum);
      edges.push({
        source: prevId,
        target: elseId,
        type: 'RACE',
        line: lineNum,
      });
      prevId = elseId;
      continue;
    }

    // for/while → process (loop body is a traced feedback)
    if (/\b(for|while)\s*\(/.test(line)) {
      const loopId = addNode(`loop_L${lineNum}`, lineNum);
      edges.push({
        source: prevId,
        target: loopId,
        type: 'PROCESS',
        line: lineNum,
      });
      prevId = loopId;
      continue;
    }

    // function/const arrow → new node
    if (
      /^(export\s+)?(async\s+)?function\s+(\w+)/.test(line) ||
      /^(export\s+)?const\s+\w+\s*=\s*(async\s+)?\(/.test(line)
    ) {
      const match =
        line.match(/function\s+(\w+)/) ?? line.match(/const\s+(\w+)/);
      const name = match?.[1] ?? `fn_L${lineNum}`;
      const fnId = addNode(name, lineNum);
      edges.push({
        source: prevId,
        target: fnId,
        type: 'PROCESS',
        line: lineNum,
      });
      prevId = fnId;
      continue;
    }
  }

  // Exit node
  const exitId = addNode('exit', lines.length);
  edges.push({
    source: prevId,
    target: exitId,
    type: 'FOLD',
    line: lines.length,
  });

  // Generate GG source
  const ggLines: string[] = [
    `// Generated from ${filename}`,
    `// ${nodes.length} nodes, ${edges.length} edges`,
    '',
  ];

  for (const node of nodes) {
    ggLines.push(`(${node.id} : ${node.label})`);
  }
  ggLines.push('');

  for (const edge of edges) {
    ggLines.push(`(${edge.source})-[${edge.type}]->(${edge.target})`);
  }

  const stats = {
    forks: edges.filter((e) => e.type === 'FORK').length,
    races: edges.filter((e) => e.type === 'RACE').length,
    folds: edges.filter((e) => e.type === 'FOLD').length,
    vents: edges.filter((e) => e.type === 'VENT').length,
    processes: edges.filter((e) => e.type === 'PROCESS').length,
    linesScanned: lines.length,
  };

  return {
    ggSource: ggLines.join('\n'),
    nodes,
    edges,
    stats,
  };
}
