// @a0n/.ignored_gnosis/src/ts-topology-slo.ts
/**
 * Gnosis Topology SLOs
 *
 * Define and enforce Service Level Objectives on topology metrics.
 * Used in CI pipelines to enforce complexity budgets.
 *
 * Example SLO config:
 *   maxBuley: 30
 *   maxRegime: 'transitional'
 *   maxBeta1: 5
 *   requireCertificate: true
 */

import * as fs from 'node:fs';
import * as path from 'node:path';
import {
  checkTypeScriptWithGnosis,
  type GnosisTypeScriptCheckResult,
} from './ts-check';
import type { GnosisSteeringRegime } from './analysis';

export interface TopologySloConfig {
  readonly maxBuley?: number;
  readonly maxWallace?: number;
  readonly maxBeta1?: number;
  readonly maxRegime?: GnosisSteeringRegime;
  readonly maxNodeCount?: number;
  readonly maxEdgeCount?: number;
  readonly requireCertificate?: boolean;
  readonly paths?: readonly string[];
  readonly excludePaths?: readonly string[];
}

export interface SloViolation {
  readonly filePath: string;
  readonly metric: string;
  readonly actual: number | string;
  readonly threshold: number | string;
  readonly message: string;
}

export interface SloCheckResult {
  readonly ok: boolean;
  readonly violations: readonly SloViolation[];
  readonly checked: number;
  readonly passed: number;
  readonly failed: number;
  readonly skipped: number;
}

const REGIME_ORDER: Record<GnosisSteeringRegime, number> = {
  laminar: 0,
  transitional: 1,
  turbulent: 2,
};

/**
 * Loads the Slo Config.
 */
export function loadSloConfig(configPath: string): TopologySloConfig {
  // polyglot:ignore RESOURCE_LEAK — readFileSync returns a string, no handle to release
  const content = fs.readFileSync(configPath, 'utf-8');
  if (configPath.endsWith('.json')) {
    return JSON.parse(content) as TopologySloConfig;
  }
  // Support TOML-like simple format
  const config: Record<string, unknown> = {};
  for (const line of content.split('\n')) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#')) continue;
    const [key, ...valueParts] = trimmed.split('=');
    const value = valueParts.join('=').trim();
    if (key && value) {
      const k = key.trim();
      if (value === 'true') config[k] = true;
      else if (value === 'false') config[k] = false;
      else if (/^\d+(\.\d+)?$/.test(value)) config[k] = Number(value);
      else config[k] = value.replace(/^['"]|['"]$/g, '');
    }
  }
  return config as unknown as TopologySloConfig;
}

/**
 * Handles the gnosis check Topology Slo workflow.
 */
export async function checkTopologySlo(
  filePath: string,
  slo: TopologySloConfig
): Promise<readonly SloViolation[]> {
  const violations: SloViolation[] = [];

  let result: GnosisTypeScriptCheckResult;
  try {
    // polyglot:ignore RESOURCE_LEAK — readFileSync returns a string, no handle to release
    const sourceText = fs.readFileSync(filePath, 'utf-8');
    result = await checkTypeScriptWithGnosis(sourceText, filePath);
  } catch {
    return []; // Unbridgeable files pass SLO by default
  }

  const { metrics } = result;

  if (slo.maxBuley !== undefined && metrics.buleyNumber > slo.maxBuley) {
    violations.push({
      filePath,
      metric: 'buleyNumber',
      actual: metrics.buleyNumber,
      threshold: slo.maxBuley,
      message: `Buley ${metrics.buleyNumber} exceeds SLO threshold ${slo.maxBuley}`,
    });
  }

  if (slo.maxWallace !== undefined && metrics.wallaceNumber > slo.maxWallace) {
    violations.push({
      filePath,
      metric: 'wallaceNumber',
      actual: metrics.wallaceNumber,
      threshold: slo.maxWallace,
      message: `Wallace ${metrics.wallaceNumber} exceeds SLO threshold ${slo.maxWallace}`,
    });
  }

  if (slo.maxBeta1 !== undefined && metrics.beta1 > slo.maxBeta1) {
    violations.push({
      filePath,
      metric: 'beta1',
      actual: metrics.beta1,
      threshold: slo.maxBeta1,
      message: `Beta-1 ${metrics.beta1} exceeds SLO threshold ${slo.maxBeta1}`,
    });
  }

  if (slo.maxRegime !== undefined) {
    const maxOrder = REGIME_ORDER[slo.maxRegime];
    const actualOrder = REGIME_ORDER[metrics.regime];
    if (actualOrder > maxOrder) {
      violations.push({
        filePath,
        metric: 'regime',
        actual: metrics.regime,
        threshold: slo.maxRegime,
        message: `Regime '${metrics.regime}' exceeds SLO maximum '${slo.maxRegime}'`,
      });
    }
  }

  if (slo.maxNodeCount !== undefined && metrics.nodeCount > slo.maxNodeCount) {
    violations.push({
      filePath,
      metric: 'nodeCount',
      actual: metrics.nodeCount,
      threshold: slo.maxNodeCount,
      message: `Node count ${metrics.nodeCount} exceeds SLO threshold ${slo.maxNodeCount}`,
    });
  }

  if (slo.maxEdgeCount !== undefined && metrics.edgeCount > slo.maxEdgeCount) {
    violations.push({
      filePath,
      metric: 'edgeCount',
      actual: metrics.edgeCount,
      threshold: slo.maxEdgeCount,
      message: `Edge count ${metrics.edgeCount} exceeds SLO threshold ${slo.maxEdgeCount}`,
    });
  }

  return violations;
}

/**
 * Handles the gnosis check Directory Slo workflow.
 */
export async function checkDirectorySlo(
  dirPath: string,
  slo: TopologySloConfig
): Promise<SloCheckResult> {
  const files = findTypeScriptFiles(dirPath, slo.paths, slo.excludePaths);
  const allViolations: SloViolation[] = [];
  let passed = 0;
  let failed = 0;
  let skipped = 0;

  for (const file of files) {
    try {
      const violations = await checkTopologySlo(file, slo);
      if (violations.length > 0) {
        allViolations.push(...violations);
        failed += 1;
      } else {
        passed += 1;
      }
    } catch {
      skipped += 1;
    }
  }

  return {
    ok: allViolations.length === 0,
    violations: allViolations,
    checked: files.length,
    passed,
    failed,
    skipped,
  };
}

function findTypeScriptFiles(
  dirPath: string,
  includePaths?: readonly string[],
  excludePaths?: readonly string[]
): string[] {
  const files: string[] = [];
  const absDir = path.resolve(dirPath);

  const walk = (dir: string): void => {
    let entries: fs.Dirent[];
    try {
      entries = fs.readdirSync(dir, { withFileTypes: true });
    } catch {
      return;
    }

    for (const entry of entries) {
      const full = path.join(dir, entry.name);
      const relative = path.relative(absDir, full);

      if (entry.isDirectory()) {
        if (
          [
            'node_modules',
            'dist',
            '.next',
            '.bun',
            'build',
            'coverage',
          ].includes(entry.name)
        ) {
          continue;
        }
        if (excludePaths?.some((p) => relative.startsWith(p))) continue;
        walk(full);
      } else if (entry.isFile()) {
        if (!full.endsWith('.ts') && !full.endsWith('.tsx')) continue;
        if (full.endsWith('.d.ts')) continue;
        if (
          full.includes('.test.') ||
          full.includes('.spec.') ||
          full.includes('__tests__')
        )
          continue;
        if (excludePaths?.some((p) => relative.startsWith(p))) continue;
        if (
          includePaths &&
          includePaths.length > 0 &&
          !includePaths.some((p) => relative.startsWith(p))
        )
          continue;
        files.push(full);
      }
    }
  };

  walk(absDir);
  return files;
}
