// @a0n/.ignored_gnosis/src/twin-paradox.ts
/**
 * Twin Paradox -- time dilation for round-trip journeys.
 *
 * Computes the age difference between a stationary twin and a
 * traveling twin who makes a round trip at relativistic speed.
 * The traveling twin follows a broken geodesic (acceleration at
 * the turnaround), so their proper time is strictly less than
 * the stationary twin's coordinate time.
 *
 * Natural units: c = 1 lpLo/pLo (one light-picolorenzo per picolorenzo).
 * Distances arrive in light-years and are converted to lpLo for
 * the computation. Time is in picolorenzos throughout.
 *
 * 1 pLo = pi days, so 1 year = 365.25 / pi pLo.
 *
 * @packageDocumentation
 */

import { lorentzGamma } from './simultaneity';
import { LIGHT_PICOLORENZO_LY } from './math';

// ─── Types ──────────────────────────────────────────────────────────

export interface TwinParadoxInput {
  readonly targetDistanceLy: number;
  readonly travelBeta: number; // fraction of c, (0, 1)
}

export interface TwinLeg {
  readonly direction: 'outbound' | 'return';
  readonly distanceLy: number;
  readonly beta: number;
  readonly gamma: number;
  readonly coordinateTimePLo: number;
  readonly properTimePLo: number;
  readonly startX: number;
  readonly endX: number;
  readonly startPLo: number;
  readonly endPLo: number;
}

export interface TwinWorldline {
  readonly label: string;
  readonly legs: readonly TwinLeg[];
  readonly totalCoordinateTimePLo: number;
  readonly totalProperTimePLo: number;
}

export interface TwinParadoxResult {
  readonly source: 'GnosisTwinParadox';
  readonly targetDistanceLy: number;
  readonly travelBeta: number;
  readonly gamma: number;
  readonly stationaryTwin: TwinWorldline;
  readonly travelingTwin: TwinWorldline;
  readonly ageDifferencePLo: number;
  readonly ageDifferenceYears: number;
  readonly timeDilationRatio: number; // traveler_proper / stationary_proper
}

// ─── Constants ──────────────────────────────────────────────────────

/** 1 year in picolorenzos. 1 pLo = pi days, so 1 year = 365.25/pi pLo. */
const PLO_PER_YEAR = 365.25 / Math.PI;

/** Convert picolorenzos to years. */
function ploToYears(pLo: number): number {
  return pLo / PLO_PER_YEAR;
}

// ─── Single Round Trip ──────────────────────────────────────────────

/**
 * Compute the twin paradox for a symmetric round trip.
 *
 * The traveling twin departs at beta, travels targetDistanceLy,
 * turns around (instantaneous acceleration), and returns at the
 * same speed. The stationary twin remains at x = 0.
 *
 * All times are in picolorenzos. Distances in light-years for
 * the interface, converted to light-picolorenzos internally.
 */
export function twinParadox(input: TwinParadoxInput): TwinParadoxResult {
  const { targetDistanceLy, travelBeta } = input;
  const gamma = lorentzGamma(travelBeta);

  // Distance in light-picolorenzos (natural units where c = 1 lpLo/pLo)
  const dLpLo = targetDistanceLy / LIGHT_PICOLORENZO_LY;

  // Outbound coordinate time: t = d / beta (in pLo)
  const tOutCoord = dLpLo / travelBeta;
  // Return coordinate time: same by symmetry
  const tRetCoord = dLpLo / travelBeta;
  // Total coordinate time
  const totalCoordTime = tOutCoord + tRetCoord;

  // Proper time for each leg: t_proper = t_coord / gamma
  const tOutProper = tOutCoord / gamma;
  const tRetProper = tRetCoord / gamma;
  const totalTravelerProper = tOutProper + tRetProper;

  // Stationary twin: proper time = coordinate time (no motion)
  const totalStationaryProper = totalCoordTime;

  // Age difference
  const ageDifferencePLo = totalStationaryProper - totalTravelerProper;

  // ─── Build legs ─────────────────────────────────────────────────

  const outboundLeg: TwinLeg = {
    direction: 'outbound',
    distanceLy: targetDistanceLy,
    beta: travelBeta,
    gamma,
    coordinateTimePLo: tOutCoord,
    properTimePLo: tOutProper,
    startX: 0,
    endX: dLpLo,
    startPLo: 0,
    endPLo: tOutCoord,
  };

  const returnLeg: TwinLeg = {
    direction: 'return',
    distanceLy: targetDistanceLy,
    beta: travelBeta,
    gamma,
    coordinateTimePLo: tRetCoord,
    properTimePLo: tRetProper,
    startX: dLpLo,
    endX: 0,
    startPLo: tOutCoord,
    endPLo: totalCoordTime,
  };

  const stationaryLeg: TwinLeg = {
    direction: 'outbound', // stays put, direction is nominal
    distanceLy: 0,
    beta: 0,
    gamma: 1,
    coordinateTimePLo: totalCoordTime,
    properTimePLo: totalStationaryProper,
    startX: 0,
    endX: 0,
    startPLo: 0,
    endPLo: totalCoordTime,
  };

  const stationaryTwin: TwinWorldline = {
    label: 'stationary',
    legs: [stationaryLeg],
    totalCoordinateTimePLo: totalCoordTime,
    totalProperTimePLo: totalStationaryProper,
  };

  const travelingTwin: TwinWorldline = {
    label: 'traveling',
    legs: [outboundLeg, returnLeg],
    totalCoordinateTimePLo: totalCoordTime,
    totalProperTimePLo: totalTravelerProper,
  };

  return {
    source: 'GnosisTwinParadox',
    targetDistanceLy,
    travelBeta,
    gamma,
    stationaryTwin,
    travelingTwin,
    ageDifferencePLo,
    ageDifferenceYears: ploToYears(ageDifferencePLo),
    timeDilationRatio: totalTravelerProper / totalStationaryProper,
  };
}

// ─── Multi-Leg Journey ──────────────────────────────────────────────

/**
 * Compute the twin paradox for a multi-leg journey.
 *
 * Each leg has its own distance and speed. The traveler chains
 * the legs sequentially: leg 1 is outbound, leg 2 may be outbound
 * or return, etc. The stationary twin sits at x = 0 for the total
 * coordinate time of all legs combined.
 *
 * This generalizes twinParadox to asymmetric and multi-segment trips.
 */
export function twinParadoxMultiLeg(
  legs: readonly { distanceLy: number; beta: number }[]
): TwinParadoxResult {
  const builtLegs: TwinLeg[] = [];
  let currentX = 0;
  let currentPLo = 0;
  let totalTravelerProper = 0;
  let totalCoordTime = 0;
  let totalDistanceLy = 0;

  for (let i = 0; i < legs.length; i++) {
    const { distanceLy, beta } = legs[i]!;
    const gamma = lorentzGamma(beta);
    const dLpLo = distanceLy / LIGHT_PICOLORENZO_LY;
    const coordTime = dLpLo / beta;
    const properTime = coordTime / gamma;

    // Alternate outbound/return for direction labeling
    const isReturn = i % 2 === 1;
    const endX = isReturn ? currentX - dLpLo : currentX + dLpLo;

    builtLegs.push({
      direction: isReturn ? 'return' : 'outbound',
      distanceLy,
      beta,
      gamma,
      coordinateTimePLo: coordTime,
      properTimePLo: properTime,
      startX: currentX,
      endX,
      startPLo: currentPLo,
      endPLo: currentPLo + coordTime,
    });

    currentX = endX;
    currentPLo += coordTime;
    totalTravelerProper += properTime;
    totalCoordTime += coordTime;
    totalDistanceLy += distanceLy;
  }

  const totalStationaryProper = totalCoordTime;

  // Use the first leg's beta and gamma for the top-level fields
  const primaryBeta = legs.length > 0 ? legs[0]!.beta : 0;
  const primaryGamma = legs.length > 0 ? lorentzGamma(legs[0]!.beta) : 1;

  const stationaryLeg: TwinLeg = {
    direction: 'outbound',
    distanceLy: 0,
    beta: 0,
    gamma: 1,
    coordinateTimePLo: totalCoordTime,
    properTimePLo: totalStationaryProper,
    startX: 0,
    endX: 0,
    startPLo: 0,
    endPLo: totalCoordTime,
  };

  const stationaryTwin: TwinWorldline = {
    label: 'stationary',
    legs: [stationaryLeg],
    totalCoordinateTimePLo: totalCoordTime,
    totalProperTimePLo: totalStationaryProper,
  };

  const travelingTwin: TwinWorldline = {
    label: 'traveling',
    legs: builtLegs,
    totalCoordinateTimePLo: totalCoordTime,
    totalProperTimePLo: totalTravelerProper,
  };

  const ageDifferencePLo = totalStationaryProper - totalTravelerProper;

  return {
    source: 'GnosisTwinParadox',
    targetDistanceLy: totalDistanceLy,
    travelBeta: primaryBeta,
    gamma: primaryGamma,
    stationaryTwin,
    travelingTwin,
    ageDifferencePLo,
    ageDifferenceYears: ploToYears(ageDifferencePLo),
    timeDilationRatio: totalTravelerProper / totalStationaryProper,
  };
}
