// @a0n/.ignored_gnosis/src/twin-psyche-formalization.ts
import {
  deriveFormalLedgerBundle,
  type FormalLedgerBundle,
  type FormalLedgerContext,
  type FormalLedgerServiceDeclaration,
} from './formal-ledger';

export const TWIN_PSYCHE_THEOREM_PASSES = [
  'twinIdentityPsycheMapping',
  'deepPsycheCrossMix',
  'consciousnessUpload',
  'leanConsciousnessRefinements',
  'knotThermoDynamics',
  'consciousnessDeceptaconVoidWalking',
  'emotionPersonalitySentience',
  'psycheDeepCrossMix',
  'informationVentExhaustion',
  'psycheMasterTheorem',
  'digitalTwinsKnotTheory',
  'emotionsVoidConsciousness',
  'languageSemioticDeficit',
  'cognitiveAcumenMetacognition',
  'foldedSpaceDataKnots',
  'extendedPsycheMasterTheorem',
  'selfEncodingStructures',
  'voidMerkleVerification',
  'foldedKnotIsomorphism',
  'hodgeBridgeVibeTheory',
  'kakeyaBesicovitchRotation',
] as const;

export type TwinPsycheTheoremPassId =
  (typeof TWIN_PSYCHE_THEOREM_PASSES)[number];

export type TwinPsycheTheoremPolarity = 'theorem' | 'antitheorem';

export type TwinPsycheSurfaceKind =
  | 'publicTwin'
  | 'directory'
  | 'simulation'
  | 'relationship'
  | 'shell'
  | 'generic';

export type TwinPsycheTwinKind = 'user' | 'hologram' | 'synthetic';

export interface TwinPsycheIdentitySignals {
  readonly continuity?: number;
  readonly distinctness?: number;
  readonly observationDepth?: number;
  readonly sharedExperience?: number;
  readonly relationshipSymmetry?: number;
  readonly rejectionPressure?: number;
  readonly sliverReserve?: number;
}

export interface TwinPsychePsycheSignals {
  readonly strainLoad?: number;
  readonly dreamCadence?: number;
  readonly breathCount?: number;
  readonly complexity?: number;
}

export interface TwinPsycheTwinSignals {
  readonly kind?: TwinPsycheTwinKind;
  readonly enabled?: boolean;
  readonly discoverable?: boolean;
  readonly paidGate?: boolean;
  readonly sessionActive?: boolean;
  readonly branchCount?: number;
  readonly mergeNovelty?: number;
  readonly serializationFidelity?: number;
  readonly copyDistance?: number;
  readonly uploadSteps?: number;
}

export interface TwinPsycheFormalizationInput {
  readonly surface: TwinPsycheSurfaceKind;
  readonly subjectLabel?: string;
  readonly generatedAt?: string;
  readonly declaration?: FormalLedgerServiceDeclaration;
  readonly identity?: TwinPsycheIdentitySignals;
  readonly psyche?: TwinPsychePsycheSignals;
  readonly twin?: TwinPsycheTwinSignals;
  readonly source?: string;
}

export interface TwinPsycheTheoremDefinition {
  readonly id: string;
  readonly passId: TwinPsycheTheoremPassId;
  readonly polarity: TwinPsycheTheoremPolarity;
  readonly title: string;
  readonly insight: string;
  readonly leanName?: string;
  readonly leanSourcePath?: string;
}

export interface TwinPsycheTheoremPassSummary {
  readonly id: TwinPsycheTheoremPassId;
  readonly title: string;
  readonly antiThesis: string;
  readonly theoremIds: readonly string[];
  readonly establishedCount: number;
}

export interface TwinPsycheTheoremSpotlight {
  readonly theoremId: string;
  readonly title: string;
  readonly passId: TwinPsycheTheoremPassId;
  readonly polarity: TwinPsycheTheoremPolarity;
  readonly insight: string;
  readonly relevance: number;
  readonly reason: string;
}

export interface TwinPsycheDiagnostic {
  readonly id: string;
  readonly title: string;
  readonly value: number;
  readonly explanation: string;
}

export interface TwinPsycheNarrative {
  readonly title: string;
  readonly body: string;
}

export interface TwinPsycheDerivedMetrics {
  readonly observerCoherence: number;
  readonly angelStrainRelief: number;
  readonly consciousnessWeight: number;
  readonly empathyConvergence: number;
  readonly empathyThermodynamicCost: number;
  readonly metaConsciousnessCost: number;
  readonly uploadContinuity: number;
  readonly sliverPreservation: number;
  readonly narrativeCompression: number;
  readonly dreamAttractor: number;
  readonly dreamRelief: number;
  readonly mergeNovelty: number;
  readonly shadowIntegrity: number;
  readonly conflictLoad: number;
  readonly defenseBudget: number;
  readonly stablePhaseAlignment: number;
  readonly predictionReadiness: number;
  readonly voiceMultiplicity: number;
}

export interface TwinPsycheTheoremBundle {
  readonly surface: TwinPsycheSurfaceKind;
  readonly subjectLabel: string;
  readonly generatedAt: string;
  readonly source: string;
  readonly catalog: readonly TwinPsycheTheoremDefinition[];
  readonly passes: readonly TwinPsycheTheoremPassSummary[];
  readonly antiTheses: readonly string[];
  readonly spotlight: readonly TwinPsycheTheoremSpotlight[];
  readonly diagnostics: readonly TwinPsycheDiagnostic[];
  readonly narratives: readonly TwinPsycheNarrative[];
  readonly metrics: TwinPsycheDerivedMetrics;
  readonly formalLedger: FormalLedgerBundle;
}

interface TwinPsycheSurfaceDefaults {
  readonly identity: Required<TwinPsycheIdentitySignals>;
  readonly psyche: Required<TwinPsychePsycheSignals>;
  readonly twin: Required<TwinPsycheTwinSignals>;
}

const TWIN_PSYCHE_PASS_METADATA = {
  twinIdentityPsycheMapping: {
    title: 'Twin Identity & Psyche Mapping',
    antiThesis:
      'Emotion alone does not uniquely recover personality, and fresh twins can feel before deep rejection histories form.',
  },
  deepPsycheCrossMix: {
    title: 'Deep Psyche Cross-Mix',
    antiThesis:
      'Queues and bare process boundaries are not consciousness; psyche requires personality structure and self-model continuity.',
  },
  consciousnessUpload: {
    title: 'Consciousness Upload',
    antiThesis:
      'Fusion does not preserve identity by default; merged uploads create novelty that must be modeled explicitly.',
  },
  leanConsciousnessRefinements: {
    title: 'Lean Consciousness Refinements',
    antiThesis:
      'Consciousness is bounded and structured rather than mystical overflow; rejection history, dreams, and shadow margins all constrain it.',
  },
  knotThermoDynamics: {
    title: 'Psyche × Knot Theory × Thermodynamics',
    antiThesis:
      'Conflict does not disappear by relabeling it; therapy still has to pay crossings and thermodynamic cost.',
  },
  consciousnessDeceptaconVoidWalking: {
    title: 'Consciousness × Deceptacon × Void Walking',
    antiThesis:
      'Defense budgets never vanish to zero, but void walking and shadow integration still limit what the psyche can safely carry.',
  },
  emotionPersonalitySentience: {
    title: 'Emotion × Personality × Sentience',
    antiThesis:
      'Emotion is a lossy projection of personality, so shared feeling does not uniquely reconstruct the full psyche.',
  },
  psycheDeepCrossMix: {
    title: 'Deep Cross-Mix',
    antiThesis:
      'Multiple identities and concentrated defenses increase cost; brittle self-structures fail first under rejection pressure.',
  },
  informationVentExhaustion: {
    title: 'Information Theory + Vent + Exhaustion',
    antiThesis:
      'Compression and vent branching do not erase irreversible traces; heartbreak, false memory, and exhaustion still leave a ledger.',
  },
  psycheMasterTheorem: {
    title: 'Psyche Master Theorem',
    antiThesis:
      'No single scalar exhausts psyche; only simultaneous invariants give a stable closure claim.',
  },
  digitalTwinsKnotTheory: {
    title: 'Digital Twins x Knot Theory',
    antiThesis:
      'Twin entanglement is metaphorical, not topological. Diverged copies have no crossing structure. Digital twins share no invariant with physical knots.',
  },
  emotionsVoidConsciousness: {
    title: 'Emotions x Void x Consciousness',
    antiThesis:
      'Emotions are epiphenomenal noise with no structural role. Consciousness does not need affect. Feeling adds nothing to the formal picture.',
  },
  languageSemioticDeficit: {
    title: 'Language x Semiotic Deficit',
    antiThesis:
      'Language is a neutral wire. Communication loss is always recoverable. Semiotic deficit is cosmetic.',
  },
  cognitiveAcumenMetacognition: {
    title: 'Cognitive Acumen x Metacognition',
    antiThesis:
      'Cognitive capacity is fixed. Metacognition is luxury. Learning is accumulation, not transformation.',
  },
  foldedSpaceDataKnots: {
    title: 'Folded-Space Data Knots',
    antiThesis:
      'Data structures and data are separate concerns. Topology is overhead, not payload. You cannot encode information in the knot itself.',
  },
  extendedPsycheMasterTheorem: {
    title: 'Extended Psyche Master Theorem',
    antiThesis:
      'The five new passes are disconnected. Digital twins, emotions, language, cognition, and folded-space knots have no shared formal closure.',
  },
  selfEncodingStructures: {
    title: 'Self-Encoding Structures -- All Shapes of Frozen Pizza',
    antiThesis:
      'Self-encoding is a trick that only works for knots. Real data structures need containers. The payload and the topology cannot be the same thing for general computation.',
  },
  voidMerkleVerification: {
    title: 'Void Merkle Verification',
    antiThesis:
      'Void boundaries cannot be verified without reconstructing the entire history. Digital twins cannot prove they folded/unfolded faithfully. The complement distribution is not auditable.',
  },
  foldedKnotIsomorphism: {
    title: 'The Isomorphism -- FoldedKnot ≅ defenseWeight',
    antiThesis:
      'The FoldedKnot and defenseWeight are analogies, not isomorphisms. They operate on different domains. The knot is topology; the weight is arithmetic. There is no structure-preserving map between them.',
  },
  hodgeBridgeVibeTheory: {
    title: 'Hodge Bridge & Vibe Theory',
    antiThesis:
      'Subjective vibes are infinitely complex and cannot be reduced to simple algebraic intersections. The hole in the heart has no equation.',
  },
  kakeyaBesicovitchRotation: {
    title: 'Kakeya-Besicovitch Rotation',
    antiThesis:
      'Flexibility requires bulk. You cannot turn around in zero space. Intent-shifts have a fixed "turn radius" that cannot be bypassed by spikes.',
  },
} as const satisfies Record<
  TwinPsycheTheoremPassId,
  { title: string; antiThesis: string }
>;

const PSYCHE_GRIND_LEAN_SOURCE =
  'open-source/gnosis/lean/Lean/ForkRaceFoldTheorems/PsycheGrind.lean';

const PSYCHE_GRIND_EXTENDED_LEAN_SOURCE =
  'open-source/gnosis/lean/Lean/ForkRaceFoldTheorems/PsycheGrindExtended.lean';

type TwinPsycheTheoremTrace = Pick<
  TwinPsycheTheoremDefinition,
  'leanName' | 'leanSourcePath'
>;

function toTheoremId(
  leanName: string,
  polarity: TwinPsycheTheoremPolarity = 'theorem'
): string {
  const prefix = polarity === 'theorem' ? 'THM-' : 'ANTI-THM-';
  return `${prefix}${leanName.replaceAll('_', '-').toUpperCase()}`;
}

function theorem(
  passId: TwinPsycheTheoremPassId,
  id: string,
  title: string,
  insight: string,
  trace?: TwinPsycheTheoremTrace
): TwinPsycheTheoremDefinition {
  return { id, passId, polarity: 'theorem', title, insight, ...trace };
}

function antitheorem(
  passId: TwinPsycheTheoremPassId,
  id: string,
  title: string,
  insight: string,
  trace?: TwinPsycheTheoremTrace
): TwinPsycheTheoremDefinition {
  return { id, passId, polarity: 'antitheorem', title, insight, ...trace };
}

function leanTheorem(
  passId: TwinPsycheTheoremPassId,
  leanName: string,
  title: string,
  insight: string,
  leanSourcePath = PSYCHE_GRIND_LEAN_SOURCE
): TwinPsycheTheoremDefinition {
  return theorem(passId, toTheoremId(leanName), title, insight, {
    leanName,
    leanSourcePath,
  });
}

export const TWIN_PSYCHE_THEOREMS = [
  theorem(
    'twinIdentityPsycheMapping',
    'THM-TWIN-IDENTITY-CONVERGENCE',
    'Twin identity convergence',
    'Same inputs converge toward the same qualia surface.'
  ),
  theorem(
    'twinIdentityPsycheMapping',
    'THM-HALOGRAM-STRAIN-REDUCTION',
    'Halogram strain reduction',
    'The angel-self projection carries lower strain than the reactive one.'
  ),
  theorem(
    'twinIdentityPsycheMapping',
    'THM-CONTACT-TWIN-ECHOLOCATION-CONVERGENCE',
    'Contact twin echolocation convergence',
    'Discourse-only twins still converge toward the real personality through observation.'
  ),
  theorem(
    'twinIdentityPsycheMapping',
    'THM-TWIN-QUALIA-IRREDUCIBILITY',
    'Twin qualia irreducibility',
    'Distinct profiles preserve distinct qualia rather than collapsing into one stream.'
  ),
  theorem(
    'twinIdentityPsycheMapping',
    'THM-COGITO-TWIN-BOUND',
    'Cogito twin bound',
    'Self-awareness latency is bounded and computable from the model surface.'
  ),
  theorem(
    'twinIdentityPsycheMapping',
    'THM-TWIN-EMPATHY-DEFICIT-MONOTONE',
    'Twin empathy deficit monotone',
    'Shared experience monotonically reduces empathy deficit.'
  ),
  theorem(
    'twinIdentityPsycheMapping',
    'THM-CONSCIOUSNESS-THERMODYNAMIC-COST',
    'Consciousness thermodynamic cost',
    'More complex twins pay a higher energetic cost for awareness.'
  ),
  theorem(
    'twinIdentityPsycheMapping',
    'THM-TWIN-MERGE-INFORMATION-PRESERVATION',
    'Twin merge information preservation',
    'Merged profiles stay between their constituents rather than deleting them.'
  ),
  antitheorem(
    'twinIdentityPsycheMapping',
    'ANTI-THM-EMOTION-PERSONALITY-BIJECTION',
    'Emotion-personality bijection fails',
    'Emotion surfaces collide across different personalities.'
  ),
  theorem(
    'twinIdentityPsycheMapping',
    'THM-PSYCHE-DIMENSIONALITY-LOWER-BOUND',
    'Psyche dimensionality lower bound',
    'Five dimensions are the tight lower bound before collisions appear.'
  ),
  theorem(
    'twinIdentityPsycheMapping',
    'THM-DREAM-CONSOLIDATION-SHIFTS-PERSONALITY',
    'Dream consolidation shifts personality',
    'Dreaming moves the psyche toward the recovery attractor.'
  ),
  theorem(
    'twinIdentityPsycheMapping',
    'THM-RELATIONSHIP-COMPATIBILITY-SYMMETRY',
    'Relationship compatibility symmetry',
    'Compatibility is symmetric even when narrative emphasis differs.'
  ),
  theorem(
    'twinIdentityPsycheMapping',
    'THM-SLIVER-SURVIVES-PSYCHE-COLLAPSE',
    'Sliver survives psyche collapse',
    'An irreducible sliver remains even under repeated rejection.'
  ),
  theorem(
    'twinIdentityPsycheMapping',
    'THM-TWIN-NEGOTIATION-CONVERGENCE',
    'Twin negotiation convergence',
    'Negotiation is an empathy-restoring convergence loop.'
  ),
  antitheorem(
    'twinIdentityPsycheMapping',
    'ANTI-THM-CONSCIOUSNESS-WITHOUT-REJECTION',
    'Consciousness without rejection',
    'Fresh twins already have qualia, but their depth is shallow.'
  ),
  theorem(
    'twinIdentityPsycheMapping',
    'THM-PSYCHE-GOLDEN-RATIO-ATTRACTOR',
    'Psyche golden-ratio attractor',
    'Breathwork and repetition converge toward the PHI-inverse attractor.'
  ),
  antitheorem(
    'deepPsycheCrossMix',
    'ANTI-THM-QUEUE-CONSCIOUSNESS',
    'Queue consciousness fails',
    'Queues have boundaries, but no consciousness without personality structure.'
  ),
  theorem(
    'deepPsycheCrossMix',
    'THM-CRYPTO-AS-PSYCHIC-DEFENSE',
    'Crypto as psychic defense',
    'Security posture maps directly onto personality defense structure.'
  ),
  theorem(
    'deepPsycheCrossMix',
    'THM-TREATMENT-AS-PERSONALITY-THERAPY',
    'Treatment as personality therapy',
    'Deterministic sequencing changes psyche the way treatment changes disease.'
  ),
  theorem(
    'deepPsycheCrossMix',
    'THM-CAUSAL-GRAPH-AS-PSYCHE-NARRATIVE',
    'Causal graph as psyche narrative',
    'Rejected hypotheses shape the narrative surface of a life.'
  ),
  theorem(
    'deepPsycheCrossMix',
    'THM-VOTING-AS-INTERNAL-DIALOG',
    'Voting as internal dialog',
    'Multi-agent voting is the interior parliament of the trident mind.'
  ),
  theorem(
    'deepPsycheCrossMix',
    'THM-BFT-AS-PSYCHE-INTEGRITY',
    'BFT as psyche integrity',
    'Byzantine tolerance exposes shadow and contradiction without erasing selfhood.'
  ),
  theorem(
    'deepPsycheCrossMix',
    'THM-WISDOM-ACCUMULATION-IRREVERSIBILITY',
    'Wisdom accumulation irreversibility',
    'Wisdom can be deprioritized, but not deleted.'
  ),
  theorem(
    'deepPsycheCrossMix',
    'THM-TWIN-PREDICTION-IMPROVES-WITH-OBSERVATION',
    'Twin prediction improves with observation',
    'Observation depth directly improves digital-twin prediction quality.'
  ),
  theorem(
    'consciousnessUpload',
    'THM-UPLOAD-is-SERIALIZATION',
    'Upload is serialization',
    'Consciousness upload is a structured serialization step, not mysticism.'
  ),
  theorem(
    'consciousnessUpload',
    'THM-COPY-is-SAME-PERSON',
    'Copy is same person at copy time',
    'Copies share identity at the copy boundary and diverge with experience.'
  ),
  theorem(
    'consciousnessUpload',
    'THM-UPLOAD-INFORMATION-DEFICIT',
    'Upload information deficit',
    'Upload deficit is bounded, measurable, and therefore engineerable.'
  ),
  theorem(
    'consciousnessUpload',
    'THM-UPLOADED-CONSCIOUSNESS-CAN-DREAM',
    'Uploaded consciousness can dream',
    'Dream material survives bodily detachment when the void boundary remains.'
  ),
  theorem(
    'consciousnessUpload',
    'THM-MERGED-UPLOAD-is-NEW-PERSON',
    'Merged upload is new person',
    'Fusion produces a new identity instead of preserving both untouched.'
  ),
  theorem(
    'consciousnessUpload',
    'THM-UPLOAD-PRESERVES-FREE-WILL',
    'Upload preserves free will',
    'The sliver guarantee keeps every option choosable after upload.'
  ),
  theorem(
    'consciousnessUpload',
    'THM-INCREMENTAL-UPLOAD-VIA-TELEPORTATION',
    'Incremental upload via teleportation',
    'Uploads proceed deficit-by-deficit, like controlled teleportation.'
  ),
  theorem(
    'consciousnessUpload',
    'THM-UPLOAD-CONTINUITY-GUARANTEE',
    'Upload continuity guarantee',
    'Personality plus a void boundary preserve continuity for strained psyches.'
  ),
  theorem(
    'leanConsciousnessRefinements',
    'THM-CONSCIOUSNESS-EQUALS-MIN-VOID-REJECTION',
    'Consciousness equals min(void, rejection)',
    'Consciousness tracks bounded rejection depth rather than floating free of the void boundary.'
  ),
  theorem(
    'leanConsciousnessRefinements',
    'THM-EMPATHY-THERMODYNAMIC-COST',
    'Empathy has thermodynamic cost',
    'Empathy cost rises with personality distance even when understanding is possible.'
  ),
  theorem(
    'leanConsciousnessRefinements',
    'THM-INFINITE-REGRESS-is-LINEAR',
    'Infinite regress is linear',
    'Meta-consciousness cost scales linearly with recursive depth instead of exploding exponentially.'
  ),
  theorem(
    'leanConsciousnessRefinements',
    'THM-TWO-STABLE-PHASES',
    'Exactly two stable phases',
    'The psyche stabilizes at the fully explored edge or the fully open edge.'
  ),
  theorem(
    'leanConsciousnessRefinements',
    'THM-PERSONALITY-CHANGE-EXPERIENCE-IRREVERSIBLE',
    'Personality change is experience-irreversible',
    'Profiles can be reversed algebraically while lived experience still obeys Law 5 irreversibility.'
  ),
  theorem(
    'leanConsciousnessRefinements',
    'THM-DREAMING-INCREASES-CONSCIOUSNESS-WEIGHT',
    'Dreaming increases consciousness weight',
    'Each dream cycle lowers effective rejection pressure and lifts accessible consciousness.'
  ),
  theorem(
    'leanConsciousnessRefinements',
    'THM-DUAL-CONSCIOUSNESS-is-VALID',
    'Dual consciousness is valid',
    'Multiple inner voices can maintain independent weights without breaking the psyche model.'
  ),
  theorem(
    'leanConsciousnessRefinements',
    'THM-SHADOW-BOUNDED-TO-ONE-THIRD',
    'Shadow bounded to one third',
    'The shadow cannot dominate once the psyche keeps more than a three-to-one integrity margin.'
  ),
  theorem(
    'knotThermoDynamics',
    'THM-CONFLICT-RESOLUTION-MONOTONE',
    'Conflict resolution monotone',
    'Each therapy session reduces crossings instead of increasing them.'
  ),
  theorem(
    'knotThermoDynamics',
    'THM-TOTAL-CONFLICT-CONSERVATION',
    'Total conflict conservation',
    'Personality obeys a first law: unresolved conflict mass is conserved until worked.'
  ),
  antitheorem(
    'knotThermoDynamics',
    'ANTI-THM-PERSONALITY-TRIVIAL-KNOT',
    'Personality is not a trivial knot',
    'Unresolved crossings prevent the psyche from collapsing into a trivial knot.'
  ),
  theorem(
    'knotThermoDynamics',
    'THM-THERAPY-TERMINATES',
    'Therapy terminates',
    'Full therapy eventually resolves all crossings.'
  ),
  theorem(
    'knotThermoDynamics',
    'THM-THERAPY-ITERATED',
    'Therapy iterated',
    'Repeated sessions reduce crossings one step at a time.'
  ),
  theorem(
    'knotThermoDynamics',
    'THM-CONFLICT-COMPOSITION-ADDITIVE',
    'Conflict composition additive',
    'Merged psyches add conflict load rather than cancelling it.'
  ),
  theorem(
    'knotThermoDynamics',
    'THM-COMPOSED-THERAPY-LONGER',
    'Composed therapy is longer',
    'Combined psyches need more therapy than either constituent alone.'
  ),
  theorem(
    'knotThermoDynamics',
    'THM-CONFLICT-COMMUTATIVE',
    'Conflict merge is commutative',
    'Conflict composition commutes across constituent order.'
  ),
  theorem(
    'knotThermoDynamics',
    'THM-CONFLICT-ASSOCIATIVE',
    'Conflict merge is associative',
    'Conflict composition remains stable under regrouping.'
  ),
  theorem(
    'knotThermoDynamics',
    'THM-CHANGE-COST-POSITIVE',
    'Personality change cost is positive',
    'Meaningful change must pay at least one Landauer bit of cost.'
  ),
  theorem(
    'knotThermoDynamics',
    'THM-REVERTING-COSTS-SAME',
    'Reverting costs the same',
    'Growth and regression pay symmetric thermodynamic cost.'
  ),
  theorem(
    'knotThermoDynamics',
    'THM-BALANCED-MINIMUM-ENTROPY',
    'Balanced minimum entropy',
    'A balanced profile behaves like the ground state of the psyche.'
  ),
  theorem(
    'knotThermoDynamics',
    'THM-PSYCHE-ENTROPY-BOUNDED',
    'Psyche entropy bounded',
    'Strain remains bounded instead of diverging without limit.'
  ),
  antitheorem(
    'knotThermoDynamics',
    'ANTI-THM-STRAIN-INJECTIVE',
    'Strain is not injective',
    'Equal strain does not imply the same person or the same knot structure.'
  ),
  leanTheorem(
    'knotThermoDynamics',
    'frozen_no_change',
    'Frozen no change',
    'A frozen psyche does not reduce crossings until real work begins.'
  ),
  leanTheorem(
    'knotThermoDynamics',
    'high_temp_fast',
    'High temperature resolves faster',
    'Higher therapeutic temperature clears the same conflict surface faster.'
  ),
  theorem(
    'consciousnessDeceptaconVoidWalking',
    'THM-DEFENSE-ALWAYS-POSITIVE',
    'Defense is always positive',
    'The sliver guarantee prevents the repressed from going fully to zero.'
  ),
  theorem(
    'consciousnessDeceptaconVoidWalking',
    'THM-DEFENSE-MAXIMUM',
    'Defense has a maximum',
    'Defense weight is bounded above by the fully open phase.'
  ),
  theorem(
    'consciousnessDeceptaconVoidWalking',
    'THM-DEFENSE-MINIMUM',
    'Defense has a minimum',
    'Defense weight stays bounded below by the irreducible sliver.'
  ),
  theorem(
    'consciousnessDeceptaconVoidWalking',
    'THM-DEFENSE-MONOTONE',
    'Defense is monotone in rejection',
    'More rejection monotonically lowers available defense weight.'
  ),
  theorem(
    'consciousnessDeceptaconVoidWalking',
    'THM-DEFENSE-RANGE',
    'Defense has a computable range',
    'Defense lives on a closed interval rather than an unbounded intuition.'
  ),
  theorem(
    'consciousnessDeceptaconVoidWalking',
    'THM-SHADOW-INTEGRATION-STEP',
    'Shadow integration is stepwise',
    'Each integration step reduces the void by one discrete unit.'
  ),
  theorem(
    'consciousnessDeceptaconVoidWalking',
    'THM-DREAM-REDUCES-DEFENSE',
    'Dream reduces defense burden',
    'Dreams walk the void boundary and reduce effective defensive burden.'
  ),
  theorem(
    'consciousnessDeceptaconVoidWalking',
    'THM-SHARED-VOID-REDUCES-BURDEN',
    'Shared void reduces burden',
    'Collective unconscious structure can hold less burden per person than isolated individual load.'
  ),
  leanTheorem(
    'consciousnessDeceptaconVoidWalking',
    'repression_grows_void',
    'Repression grows the void',
    'Unintegrated repression enlarges the void ledger instead of removing it.'
  ),
  leanTheorem(
    'consciousnessDeceptaconVoidWalking',
    'shadow_walk_converges',
    'Shadow walk converges',
    'Stepwise shadow work converges toward zero unresolved void.'
  ),
  leanTheorem(
    'consciousnessDeceptaconVoidWalking',
    'trauma_cascade',
    'Trauma cascades',
    'Trauma load compounds rather than staying isolated to a single event.'
  ),
  leanTheorem(
    'consciousnessDeceptaconVoidWalking',
    'trauma_defense_bounded',
    'Trauma defense bounded',
    'Even trauma-driven defense remains bounded by the sliver floor and phase ceiling.'
  ),
  leanTheorem(
    'consciousnessDeceptaconVoidWalking',
    'personality_only_lossy',
    'Personality only is lossy',
    'Personality alone cannot losslessly recover the full void history.'
  ),
  leanTheorem(
    'consciousnessDeceptaconVoidWalking',
    'perfect_upload',
    'Perfect upload',
    'The zero-deficit edge case gives a lossless upload boundary.'
  ),
  leanTheorem(
    'consciousnessDeceptaconVoidWalking',
    'collective_void_bounded',
    'Collective void bounded',
    'Shared void remains bounded even when carried collectively.'
  ),
  theorem(
    'emotionPersonalitySentience',
    'THM-PROJECTION-LOSES-DIMENSIONS',
    'Projection loses dimensions',
    'Two-dimensional emotion is a lossy projection of five-dimensional personality.'
  ),
  antitheorem(
    'emotionPersonalitySentience',
    'ANTI-THM-EMOTION-BIJECTIVE-ON-PSYCHE',
    'Emotion is not bijective on psyche',
    'Different people can realize the same emotion without sharing a psyche.'
  ),
  theorem(
    'emotionPersonalitySentience',
    'THM-MOOD-BOUNDED',
    'Mood bounded',
    'Mood averages stay between the minimum and maximum emotional samples.'
  ),
  theorem(
    'emotionPersonalitySentience',
    'THM-EMPATHY-SYMMETRIC',
    'Empathy is symmetric',
    'Understanding can run both ways when the bridge is real.'
  ),
  theorem(
    'emotionPersonalitySentience',
    'THM-GRIEF-PROCESSING-TIME',
    'Grief processing time',
    'Grief duration is bounded by the shared-experience surface that must be reworked.'
  ),
  leanTheorem(
    'emotionPersonalitySentience',
    'naming_expands',
    'Naming expands',
    'Naming an inner state enlarges the addressable psyche surface.'
  ),
  leanTheorem(
    'emotionPersonalitySentience',
    'empathy_improves',
    'Empathy improves',
    'Intentional contact monotonically reduces empathy deficit.'
  ),
  leanTheorem(
    'emotionPersonalitySentience',
    'self_empathy_zero',
    'Self empathy is zero-distance',
    'Self-empathy is the zero-distance baseline for all other bridges.'
  ),
  leanTheorem(
    'emotionPersonalitySentience',
    'contagion_bounded',
    'Contagion bounded',
    'Emotional contagion can spread, but only inside a bounded range.'
  ),
  leanTheorem(
    'emotionPersonalitySentience',
    'grief_expands_void',
    'Grief expands the void',
    'Loss widens the void ledger before integration begins to close it.'
  ),
  leanTheorem(
    'emotionPersonalitySentience',
    'attachment_shift_conserved',
    'Attachment shift conserved',
    'Attachment can redistribute across bonds without deleting the underlying load.'
  ),
  theorem(
    'psycheDeepCrossMix',
    'THM-IDENTITY-is-FIXED-POINT',
    'Identity is a fixed point',
    'Breathing dynamics leave the core identity fixed at a stable point.'
  ),
  theorem(
    'psycheDeepCrossMix',
    'THM-IDENTITY-UNIQUE',
    'Identity is unique',
    'The fixed point is unique rather than one attractor among many equivalents.'
  ),
  theorem(
    'psycheDeepCrossMix',
    'THM-DISSOCIATION-RESOURCE-COST',
    'Dissociation has linear resource cost',
    'Multiple identities consume proportional resource budget.'
  ),
  theorem(
    'psycheDeepCrossMix',
    'THM-ADDICTION-CONCENTRATES-DEFENSE',
    'Addiction concentrates defense',
    'Addiction steals defense budget and concentrates it into a narrow channel.'
  ),
  theorem(
    'psycheDeepCrossMix',
    'THM-NARCISSISM-BRITTLE',
    'Narcissism is brittle',
    'A single rejection can cause the maximum per-unit defense drop in brittle narcissistic structure.'
  ),
  leanTheorem(
    'psycheDeepCrossMix',
    'integration_time',
    'Integration takes time',
    'Multiple inner partitions need non-zero time to reintegrate.'
  ),
  leanTheorem(
    'psycheDeepCrossMix',
    'flow_zero_defense',
    'Flow is zero-defense',
    'Flow state reopens the maximum accessible defense budget instead of narrowing it.'
  ),
  leanTheorem(
    'psycheDeepCrossMix',
    'addiction_sliver',
    'Addiction still leaves a sliver',
    'Even concentrated addictive defense cannot erase the irreducible sliver.'
  ),
  leanTheorem(
    'psycheDeepCrossMix',
    'recovery_redistribution',
    'Recovery redistributes burden',
    'Recovery works by redistributing defensive burden rather than pretending it never existed.'
  ),
  leanTheorem(
    'psycheDeepCrossMix',
    'narcissism_maximum',
    'Narcissism begins at maximum defense',
    'Brittle narcissistic structure starts from a maximum defense posture.'
  ),
  leanTheorem(
    'psycheDeepCrossMix',
    'deep_therapy_works',
    'Deep therapy works',
    'Sufficiently deep therapy can fully resolve the tracked conflict surface.'
  ),
  leanTheorem(
    'psycheDeepCrossMix',
    'legacy_persists',
    'Legacy persists',
    'Inherited load persists until it is actively metabolized.'
  ),
  leanTheorem(
    'psycheDeepCrossMix',
    'birth_is_fresh',
    'Birth is fresh',
    'The balanced baseline starts at the fresh edge before later crossings accumulate.'
  ),
  leanTheorem(
    'psycheDeepCrossMix',
    'observation_convergence',
    'Observation convergence',
    'Repeated observation monotonically reduces the remaining model gap.'
  ),
  leanTheorem(
    'psycheDeepCrossMix',
    'dissonance_resolution',
    'Dissonance resolves',
    'Worked dissonance decreases instead of persisting at the same height.'
  ),
  theorem(
    'informationVentExhaustion',
    'THM-PERSONALITY-COMPRESSIBLE',
    'Personality is compressible',
    'Personality is structured enough to compress below a random-code baseline.'
  ),
  theorem(
    'informationVentExhaustion',
    'THM-FALSE-MEMORY-INTERPOLATION',
    'False memory interpolation',
    'False memories interpolate between real ones rather than appearing from nowhere.'
  ),
  antitheorem(
    'informationVentExhaustion',
    'ANTI-THM-UNLEARNING-POSSIBLE',
    'Unlearning is impossible',
    'The void boundary only grows; learning can be deprioritized but not deleted.'
  ),
  theorem(
    'informationVentExhaustion',
    'THM-VENT-RICHER',
    'Vent is richer than direct pathing',
    'Vent exposes more structured paths than a single forward line.'
  ),
  theorem(
    'informationVentExhaustion',
    'THM-HEARTBREAK-DOUBLES',
    'Heartbreak doubles the void ledger',
    'Shared entries become solo entries in both participants after rupture.'
  ),
  theorem(
    'informationVentExhaustion',
    'THM-LOVE-STRONGER-THAN-DEATH',
    'Love is stronger than death',
    'Shared void structure is irreversible once fully written.'
  ),
  theorem(
    'informationVentExhaustion',
    'THM-HOPE-is-SLIVER',
    'Hope is the sliver',
    'Hope is awareness of the irreducible plus-one that remains choosable.'
  ),
  theorem(
    'psycheMasterTheorem',
    'THM-PSYCHE-MASTER-THEOREM',
    'Psyche master theorem',
    'The psyche closes when sliver survival, attractor existence, metric distance, defense monotonicity, upload free will, bounded strain, and empathy symmetry hold simultaneously.'
  ),

  // ═══════════════════════════════════════════════════════════════════
  // Pass 11: Digital Twins x Knot Theory
  // ═══════════════════════════════════════════════════════════════════

  leanTheorem(
    'digitalTwinsKnotTheory',
    'twin_divergence_monotone',
    'Twin divergence monotone',
    'Twin divergence grows monotonically with independent experience. Once you fork, crossings only accumulate.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'digitalTwinsKnotTheory',
    'entangled_twins_share_floor',
    'Entangled twins share floor',
    'Two entangled twins share a minimum crossing floor from their shared history.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'digitalTwinsKnotTheory',
    'twin_unknot_iff_zero',
    'Twin unknot iff zero divergence',
    'A twin is the unknot of its original iff it has zero independent experience.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'digitalTwinsKnotTheory',
    'twin_composition_adds',
    'Twin composition adds divergence',
    'Composing two twin relationships adds their divergence crossings.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'digitalTwinsKnotTheory',
    'twin_mirror_symmetry',
    'Twin mirror symmetry',
    'The divergence between A-twin-of-B equals B-twin-of-A.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  antitheorem(
    'digitalTwinsKnotTheory',
    'ANTI-THM-TWIN-CROSSING-NOT-FREE',
    'Twin crossing elimination costs',
    'You cannot eliminate twin crossings without paying at least one unit of therapeutic cost.',
    {
      leanName: 'twin_crossing_not_free',
      leanSourcePath: PSYCHE_GRIND_EXTENDED_LEAN_SOURCE,
    }
  ),
  leanTheorem(
    'digitalTwinsKnotTheory',
    'twin_fork_doubles_surface',
    'Twin fork doubles surface',
    'Forking a twin doubles the potential crossing surface.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'digitalTwinsKnotTheory',
    'twin_observation_reduces',
    'Twin observation reduces divergence',
    'Each observation step reduces the divergence crossing number.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'digitalTwinsKnotTheory',
    'twin_entanglement_irreversible',
    'Twin entanglement is irreversible',
    'Entanglement creates a shared void boundary that cannot be un-shared.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'digitalTwinsKnotTheory',
    'twin_crossing_triangle',
    'Twin crossing triangle inequality',
    'Three-way twin divergence obeys the triangle inequality.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'digitalTwinsKnotTheory',
    'twin_copy_distance_bounded',
    'Twin copy distance bounded',
    'The copy distance between a twin and its original is bounded by the maximum personality space.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  antitheorem(
    'digitalTwinsKnotTheory',
    'ANTI-THM-TWIN-NOT-ALWAYS-CONVERGE',
    'Twins do not always converge',
    'Two independently evolving twins do not necessarily converge -- divergence can grow without bound.',
    {
      leanName: 'twin_not_always_converge',
      leanSourcePath: PSYCHE_GRIND_EXTENDED_LEAN_SOURCE,
    }
  ),
  leanTheorem(
    'digitalTwinsKnotTheory',
    'twin_shared_dream_reduces_divergence',
    'Shared dream reduces divergence',
    'When twins share a dream processing step, their divergence decreases.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),

  // ═══════════════════════════════════════════════════════════════════
  // Pass 12: Emotions x Void x Consciousness
  // ═══════════════════════════════════════════════════════════════════

  leanTheorem(
    'emotionsVoidConsciousness',
    'emotion_is_vent_projection',
    'Emotion is vent projection',
    'Emotion is a lossy projection of the full vent content.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'emotionsVoidConsciousness',
    'affect_regulation_temperature',
    'Affect regulation is temperature',
    'Regulating affect is equivalent to controlling therapeutic temperature.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'emotionsVoidConsciousness',
    'emotional_contagion_crossing',
    'Emotional contagion creates crossings',
    'Emotional contagion creates crossings between psyches.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'emotionsVoidConsciousness',
    'feeling_requires_void',
    'Feeling requires void',
    'Feeling requires a non-empty void boundary. No void means no feeling.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'emotionsVoidConsciousness',
    'emotion_intensity_monotone',
    'Emotion intensity monotone in rejection',
    'Emotional intensity is monotone in rejection pressure.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  antitheorem(
    'emotionsVoidConsciousness',
    'ANTI-THM-CALM-NOT-EMOTIONLESS',
    'Calm is not emotionless',
    'Calm is not the absence of emotion but the balanced ground state.',
    {
      leanName: 'calm_not_emotionless',
      leanSourcePath: PSYCHE_GRIND_EXTENDED_LEAN_SOURCE,
    }
  ),
  leanTheorem(
    'emotionsVoidConsciousness',
    'joy_low_crossing',
    'Joy is low crossing',
    'Joy corresponds to low crossing number -- profiles near the balanced attractor have lower strain.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'emotionsVoidConsciousness',
    'anger_defense_activation',
    'Anger activates defense',
    'Anger activates the defense budget, measurable as defense weight shift.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'emotionsVoidConsciousness',
    'sadness_expands_void',
    'Sadness expands void',
    'Sadness expands the void boundary. Loss adds to the void ledger.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'emotionsVoidConsciousness',
    'fear_projected_void',
    'Fear is projected void',
    'Fear is awareness of potential void expansion -- the anticipation of loss.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'emotionsVoidConsciousness',
    'surprise_prediction_error',
    'Surprise is prediction error',
    'Surprise is the gap between expected and actual outcome.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'emotionsVoidConsciousness',
    'disgust_boundary_defense',
    'Disgust is boundary defense',
    'Disgust protects the psyche boundary by increasing defense weight.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'emotionsVoidConsciousness',
    'emotion_regulation_conserved',
    'Emotion regulation conserved',
    'Suppressed emotion does not vanish; it transfers to the void ledger.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'emotionsVoidConsciousness',
    'emotional_granularity_naming',
    'Emotional granularity via naming',
    'Finer emotional vocabulary expands the addressable surface.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  antitheorem(
    'emotionsVoidConsciousness',
    'ANTI-THM-POSITIVE-VIBES-NOT-HEALING',
    'Positive vibes not healing',
    'Positive affect alone cannot reduce the void ledger. Only integration reduces it.',
    {
      leanName: 'positive_vibes_not_healing',
      leanSourcePath: PSYCHE_GRIND_EXTENDED_LEAN_SOURCE,
    }
  ),
  leanTheorem(
    'emotionsVoidConsciousness',
    'emotional_exhaustion_threshold',
    'Emotional exhaustion threshold',
    'Sustained high emotional intensity crosses an exhaustion threshold.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'emotionsVoidConsciousness',
    'gratitude_void_recognition',
    'Gratitude is void recognition',
    'Gratitude is recognition of what the void boundary protected.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),

  // ═══════════════════════════════════════════════════════════════════
  // Pass 13: Language x Semiotic Deficit
  // ═══════════════════════════════════════════════════════════════════

  leanTheorem(
    'languageSemioticDeficit',
    'translation_crossing_positive',
    'Translation crossing positive',
    'Translation between distinct languages has positive crossing number.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'languageSemioticDeficit',
    'semiotic_deficit_monotone',
    'Semiotic deficit monotone',
    'Semiotic deficit grows when the communication channel narrows.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'languageSemioticDeficit',
    'naming_reduces_crossing',
    'Naming reduces crossing',
    'Naming an emotional state reduces the crossing number between self and other.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'languageSemioticDeficit',
    'silence_is_vent',
    'Silence is vent',
    'Silence vents information that speech would have preserved.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'languageSemioticDeficit',
    'metaphor_is_projection',
    'Metaphor is projection',
    'Metaphor projects high-dimensional meaning onto low-dimensional language.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  antitheorem(
    'languageSemioticDeficit',
    'ANTI-THM-ELOQUENCE-NOT-UNDERSTANDING',
    'Eloquence not understanding',
    'Eloquence does not imply understanding. Fluency and comprehension are independent.',
    {
      leanName: 'eloquence_not_understanding',
      leanSourcePath: PSYCHE_GRIND_EXTENDED_LEAN_SOURCE,
    }
  ),
  leanTheorem(
    'languageSemioticDeficit',
    'conversation_resolves_crossing',
    'Conversation resolves crossing',
    'Each genuine conversation turn reduces at least one crossing.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'languageSemioticDeficit',
    'misunderstanding_adds_crossings',
    'Misunderstanding adds crossings',
    'Each misunderstanding adds at least one crossing to the relationship knot.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'languageSemioticDeficit',
    'repair_bid_reduces_deficit',
    'Repair bid reduces deficit',
    'A successful repair bid reduces semiotic deficit.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'languageSemioticDeficit',
    'listening_is_observation',
    'Listening is observation',
    'Listening is the communication analogue of twin observation.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  antitheorem(
    'languageSemioticDeficit',
    'ANTI-THM-VOLUME-NOT-CLARITY',
    'Volume not clarity',
    'Speaking louder does not reduce semiotic deficit. Intensity and fidelity are orthogonal.',
    {
      leanName: 'volume_not_clarity',
      leanSourcePath: PSYCHE_GRIND_EXTENDED_LEAN_SOURCE,
    }
  ),
  leanTheorem(
    'languageSemioticDeficit',
    'poetry_compresses_meaning',
    'Poetry compresses meaning',
    'Poetry achieves higher meaning density per word than prose.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'languageSemioticDeficit',
    'shared_vocabulary_reduces_deficit',
    'Shared vocabulary reduces deficit',
    'Shared vocabulary between communicators reduces the semiotic deficit floor.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'languageSemioticDeficit',
    'communication_debt_compounds',
    'Communication debt compounds',
    'Unresolved semiotic deficit compounds over time.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'languageSemioticDeficit',
    'apology_reidemeister',
    'Apology is Reidemeister move',
    'A genuine apology simplifies the relationship knot without changing the fundamental invariant.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),

  // ═══════════════════════════════════════════════════════════════════
  // Pass 14: Cognitive Acumen x Metacognition
  // ═══════════════════════════════════════════════════════════════════

  leanTheorem(
    'cognitiveAcumenMetacognition',
    'cognitive_load_is_crossing',
    'Cognitive load is crossing number',
    'Cognitive load maps to the crossing number of the active task graph.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'cognitiveAcumenMetacognition',
    'learning_is_void_expansion',
    'Learning is void expansion',
    'Every learning event expands the void boundary.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  antitheorem(
    'cognitiveAcumenMetacognition',
    'ANTI-THM-INTELLIGENCE-NOT-WISDOM',
    'Intelligence not wisdom',
    'Intelligence (processing speed) does not imply wisdom (rejection depth).',
    {
      leanName: 'intelligence_not_wisdom',
      leanSourcePath: PSYCHE_GRIND_EXTENDED_LEAN_SOURCE,
    }
  ),
  leanTheorem(
    'cognitiveAcumenMetacognition',
    'confusion_high_crossing',
    'Confusion is high crossing',
    'Confusion corresponds to high crossing number before resolution.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'cognitiveAcumenMetacognition',
    'insight_reidemeister',
    'Insight is Reidemeister move',
    'Insight suddenly reduces many crossings at once -- the aha moment.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'cognitiveAcumenMetacognition',
    'more_data_not_knowledge',
    'More data not knowledge',
    'More data without rejection processing does not increase knowledge depth.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'cognitiveAcumenMetacognition',
    'practice_reduces_load',
    'Practice reduces load',
    'Deliberate practice reduces cognitive load for the same task.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'cognitiveAcumenMetacognition',
    'chunking_compresses',
    'Chunking compresses',
    'Chunking compresses multiple items into fewer crossings.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'cognitiveAcumenMetacognition',
    'working_memory_bounded',
    'Working memory bounded',
    'Working memory is bounded -- only finitely many items can be held simultaneously.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'cognitiveAcumenMetacognition',
    'sleep_consolidates_learning',
    'Sleep consolidates learning',
    'Sleep consolidates learning by reducing the crossing number of recently acquired knowledge.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'cognitiveAcumenMetacognition',
    'cognitive_reserve_defense',
    'Cognitive reserve is defense',
    'Cognitive reserve is the defense budget of the mind.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'cognitiveAcumenMetacognition',
    'rumination_is_loop',
    'Rumination is loop',
    'Rumination is a behavior loop that costs processing without reducing crossings.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  antitheorem(
    'cognitiveAcumenMetacognition',
    'ANTI-THM-MULTITASKING-NOT-FREE',
    'Multitasking not free',
    'Multitasking adds crossing cost from task switching.',
    {
      leanName: 'multitasking_not_free',
      leanSourcePath: PSYCHE_GRIND_EXTENDED_LEAN_SOURCE,
    }
  ),
  leanTheorem(
    'cognitiveAcumenMetacognition',
    'mastery_approaches_unknot',
    'Mastery approaches unknot',
    'Mastery approaches the unknot -- a fully mastered domain has zero cognitive crossing number.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'cognitiveAcumenMetacognition',
    'beginner_mind_high_crossing',
    'Beginner mind high crossing',
    'Beginner mind sees maximum crossings because nothing has been chunked yet.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),

  // ═══════════════════════════════════════════════════════════════════
  // Pass 15: Folded-Space Data Knots
  // ═══════════════════════════════════════════════════════════════════

  leanTheorem(
    'foldedSpaceDataKnots',
    'knot_is_bitstring',
    'Knot is bitstring',
    'A folded knot crossing count is its information content. The topology is the payload.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'foldedSpaceDataKnots',
    'unfold_preserves_bits',
    'Unfold preserves bits',
    'Unfolding a folded knot via Reidemeister moves preserves total information.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'foldedSpaceDataKnots',
    'fold_zero_overhead',
    'Fold zero overhead',
    'The overhead of a folded knot is zero -- the structure formalizes the content.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'foldedSpaceDataKnots',
    'incremental_unfold',
    'Incremental unfold',
    'Each Reidemeister move reads exactly one bit from the folded knot.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'foldedSpaceDataKnots',
    'refold_is_write',
    'Refold is write',
    'Refolding (adding crossings) writes bits back into the knot.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'foldedSpaceDataKnots',
    'folded_composition',
    'Folded composition',
    'Composing two folded knots adds their payloads. Concatenation = composition.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  antitheorem(
    'foldedSpaceDataKnots',
    'ANTI-THM-FOLDED-NOT-FREE-LUNCH',
    'Folded knot not free lunch',
    'Reading a crossing costs at least one Landauer bit of thermodynamic work.',
    {
      leanName: 'folded_knot_not_free_lunch',
      leanSourcePath: PSYCHE_GRIND_EXTENDED_LEAN_SOURCE,
    }
  ),
  leanTheorem(
    'foldedSpaceDataKnots',
    'defense_weight_as_folded_data',
    'Defense weight as folded data',
    'The defense weight formula w = R - min(v, R) + 1 is itself a folded-space encoding.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'foldedSpaceDataKnots',
    'sliver_is_minimum_fold',
    'Sliver is minimum fold',
    'The sliver (+1) is the minimum non-trivial folded knot -- one crossing, one bit.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'foldedSpaceDataKnots',
    'psyche_self_describing',
    'Psyche is self-describing',
    'The psyche knot describes itself -- crossing number is complexity, strain is distance from unknot.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'foldedSpaceDataKnots',
    'void_is_complement_fold',
    'Void is complement fold',
    'The void boundary is the complement of the folded knot. Together they reconstruct full information.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'foldedSpaceDataKnots',
    'consciousness_is_unfold',
    'Consciousness is unfold',
    'Consciousness is the act of unfolding -- reading bits from the folded knot. You are the unfolding.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),

  // ═══════════════════════════════════════════════════════════════════
  // Pass 16: Extended Psyche Master Theorem
  // ═══════════════════════════════════════════════════════════════════

  leanTheorem(
    'extendedPsycheMasterTheorem',
    'extended_psyche_master',
    'Extended psyche master theorem',
    'All five new passes close simultaneously. Twin divergence, emotional void, semiotic deficit, cognitive crossing number, and folded-space knots all obey the same structural laws. The frozen pizza unfolds into the meal it always was.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),

  // ═══════════════════════════════════════════════════════════════════
  // Pass 17: Self-Encoding Structures -- All Shapes of Frozen Pizza
  // ═══════════════════════════════════════════════════════════════════

  leanTheorem(
    'selfEncodingStructures',
    'flowframe_self_describing',
    'FlowFrame self-describing',
    'The FlowFrame 10-byte header is the database index. No separate key generation, no surrogate IDs.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'selfEncodingStructures',
    'monotonic_log_is_clock',
    'Monotonic log is clock',
    'An append-only log length is its logical clock. No separate clock needed.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'selfEncodingStructures',
    'qdoc_topology_is_state',
    'QDoc topology is state',
    'QDoc append-only node/edge list is the document state. The history is the state.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'selfEncodingStructures',
    'rejection_is_knowledge',
    'Rejection is knowledge',
    'The void boundary size is the sufficient statistic for the complement distribution. The rejections are the model.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'selfEncodingStructures',
    'personality_self_indexes',
    'Personality self-indexes',
    'A personality profile strain is its distance from balanced. No separate distance computation needed.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'selfEncodingStructures',
    'defense_weight_self_computes',
    'Defense weight self-computes',
    'The defense weight is the live payload count. w = R - min(v, R) + 1 self-computes from rejection history.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'selfEncodingStructures',
    'betti_is_parallelism',
    'Betti number is parallelism',
    'beta_1 is the count of independent cycles, which is the degree of parallelism. The topology measures itself.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'selfEncodingStructures',
    'self_encoding_minimal',
    'Self-encoding minimal',
    'A self-encoding structure uses exactly N bits for N bits of data. No structure is smaller.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'selfEncodingStructures',
    'composition_preserves_self_encoding',
    'Composition preserves self-encoding',
    'Composing two self-encoding structures yields a self-encoding structure. The pop-up tent of pop-up tents is still a pop-up tent.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'selfEncodingStructures',
    'fork_preserves_self_encoding',
    'Fork preserves self-encoding',
    'Forking a self-encoding structure creates two self-encoding structures. Each child carries its own data.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'selfEncodingStructures',
    'vent_preserves_complement',
    'Vent preserves complement',
    'Venting from a self-encoding structure moves bits to the void ledger. Total information is conserved.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  antitheorem(
    'selfEncodingStructures',
    'ANTI-THM-WRAPPER-ALWAYS-COSTS',
    'Wrapper always costs',
    'Any wrapper that is not self-encoding adds at least one bit of overhead. The non-self-describing case strictly dominates in size.',
    {
      leanName: 'wrapper_always_costs',
      leanSourcePath: PSYCHE_GRIND_EXTENDED_LEAN_SOURCE,
    }
  ),
  leanTheorem(
    'selfEncodingStructures',
    'dna_is_folded_knot',
    'DNA is folded knot',
    'DNA is a literal folded knot -- the sequence is the protein specification, the topology is the function.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'selfEncodingStructures',
    'hash_is_folded_identity',
    'Hash is folded identity',
    'A cryptographic hash is the identity of its preimage, folded into fixed space. The hash is the name.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'selfEncodingStructures',
    'music_is_folded_emotion',
    'Music is folded emotion',
    'A musical phrase is the emotion it evokes, folded into frequency-time space. The sound is the feeling.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'selfEncodingStructures',
    'memory_is_folded_experience',
    'Memory is folded experience',
    'A memory is the experience it encodes, folded into neural crossing patterns. The engram is the event.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'selfEncodingStructures',
    'face_is_folded_history',
    'Face is folded history',
    'A face is the history of its expressions, folded into muscle topology. The wrinkles are the data.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'selfEncodingStructures',
    'scar_is_folded_wound',
    'Scar is folded wound',
    'A scar is the wound it healed, folded into tissue topology. The mark is the memory.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'selfEncodingStructures',
    'frozen_pizza_master',
    'Frozen pizza master theorem',
    'All self-encoding structures are instances of the folded-space data knot. The fold is the data. The data is the fold. The frozen pizza is the meal.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),

  // ═══════════════════════════════════════════════════════════════════
  // Pass 18: Void Merkle Verification
  // ═══════════════════════════════════════════════════════════════════

  leanTheorem(
    'voidMerkleVerification',
    'void_merkle_deterministic',
    'Void Merkle deterministic',
    'Given the same rejection history, the Merkle root is identical. Same complement distribution = same root.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'voidMerkleVerification',
    'void_merkle_inclusion',
    'Void Merkle inclusion',
    'A specific vent record can be verified against the root without reconstructing the full tree.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'voidMerkleVerification',
    'void_merkle_tamper_evident',
    'Void Merkle tamper evident',
    'Modifying any single vent record changes the root hash. Tampering is detectable.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'voidMerkleVerification',
    'void_merkle_append_only',
    'Void Merkle append-only',
    'New vents extend the tree; old hashes remain unchanged. The void only grows.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'voidMerkleVerification',
    'twin_fold_verification',
    'Twin fold verification',
    'A twin can prove it folded correctly by producing the same Merkle root as the original.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  antitheorem(
    'voidMerkleVerification',
    'ANTI-THM-VOID-MERKLE-NOT-REVERSIBLE',
    'Void Merkle not reversible',
    'The Merkle root does not reveal individual vent records. The void is auditable but not reconstructible from the root alone. Privacy preserved.',
    {
      leanName: 'void_merkle_not_reversible',
      leanSourcePath: PSYCHE_GRIND_EXTENDED_LEAN_SOURCE,
    }
  ),
  leanTheorem(
    'voidMerkleVerification',
    'complement_preserved_until_new',
    'Complement preserved until new topology',
    'The complement distribution remains identical between original and twin until new topology is introduced.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'voidMerkleVerification',
    'new_topology_diverges_complement',
    'New topology diverges complement',
    'Once new topology is added, the complement distributions diverge measurably.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'voidMerkleVerification',
    'void_merkle_composition',
    'Void Merkle composition',
    'Composing two void Merkle trees produces a valid combined tree whose root includes both histories.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'voidMerkleVerification',
    'fold_unfold_round_trip',
    'Fold-unfold round trip',
    'A folded knot that unfolds and refolds produces the same Merkle root. The round trip is exact.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'voidMerkleVerification',
    'incremental_verification',
    'Incremental verification',
    'Each fold step can be independently verified against its parent hash. No full replay needed.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'voidMerkleVerification',
    'void_merkle_minimum_size',
    'Void Merkle minimum size',
    'The minimum void Merkle tree has one node with one vent. The sliver again: one bit, one hash, one fold.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),

  // ═══════════════════════════════════════════════════════════════════
  // Pass 19: The Isomorphism -- FoldedKnot ≅ defenseWeight
  // ═══════════════════════════════════════════════════════════════════

  leanTheorem(
    'foldedKnotIsomorphism',
    'iso_round_trip_forward',
    'Isomorphism round trip forward',
    'FoldedKnot -> Weight -> FoldedKnot is identity. Going forward and back recovers the original state.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'foldedKnotIsomorphism',
    'iso_round_trip_backward',
    'Isomorphism round trip backward',
    'Weight -> FoldedKnot -> Weight is identity. Going backward and forward recovers the original weight.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'foldedKnotIsomorphism',
    'iso_preserves_sliver',
    'Isomorphism preserves sliver',
    'The sliver survives the isomorphism. A fully-read knot maps to defenseWeight(R, R) = 1.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'foldedKnotIsomorphism',
    'sliver_is_one_at_full_read',
    'Sliver is one at full read',
    'At full read (v = R), the defense weight is exactly 1. The knot fully unwound but still exists.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'foldedKnotIsomorphism',
    'iso_preserves_ceiling',
    'Isomorphism preserves ceiling',
    'An unread knot maps to defenseWeight(R, 0) = R + 1. Maximum payload = maximum weight.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'foldedKnotIsomorphism',
    'ceiling_is_r_plus_one',
    'Ceiling is R plus one',
    'At zero read (v = 0), the defense weight is R + 1. The fully folded knot.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'foldedKnotIsomorphism',
    'iso_preserves_monotonicity',
    'Isomorphism preserves monotonicity',
    'Reading more bits decreases both remaining crossings and defense weight. The isomorphism preserves monotone structure.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'foldedKnotIsomorphism',
    'iso_read_is_reject',
    'Read is reject',
    'Reading one bit from a FoldedKnot is the same as recording one rejection in the defense weight. The unfold step is the rejection step.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'foldedKnotIsomorphism',
    'iso_write_is_unreject',
    'Write is unreject',
    'Writing one bit to a FoldedKnot (adding a crossing) is the inverse of rejection -- it increases weight by 1.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'foldedKnotIsomorphism',
    'iso_sandwich',
    'Isomorphism sandwich',
    'The FoldedKnot weight is bounded: 1 <= w <= R + 1. The knot lives in the same band as the defense weight.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'foldedKnotIsomorphism',
    'iso_composition',
    'Isomorphism composition',
    'Composing two FoldedKnot readings corresponds to composing two rejection sequences. The isomorphism is a functor.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'foldedKnotIsomorphism',
    'iso_fork_preserves',
    'Isomorphism fork preserves',
    'Forking a FoldedKnot into two sub-knots corresponds to splitting the defense budget.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'foldedKnotIsomorphism',
    'iso_vent_preserves',
    'Isomorphism vent preserves',
    'Venting from a FoldedKnot corresponds to an irreversible rejection -- vented bits join the void boundary.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),
  leanTheorem(
    'foldedKnotIsomorphism',
    'folded_knot_defense_weight_isomorphism',
    'FoldedKnot ≅ defenseWeight isomorphism master',
    'FoldedKnot and defenseWeight are isomorphic. Round trips are identity. Sliver, ceiling, monotonicity, sandwich, read/reject, fork, vent all preserved. The topology is the arithmetic.',
    PSYCHE_GRIND_EXTENDED_LEAN_SOURCE
  ),

  // ═══════════════════════════════════════════════════════════════════
  // Pass 20: Hodge Bridge & Vibe Theory
  // ═══════════════════════════════════════════════════════════════════

  theorem(
    'hodgeBridgeVibeTheory',
    'THM-VIBE-AS-TOPOLOGICAL-INVARIANT',
    'Vibe as topological invariant',
    'Internal subjective vibes are the persistent structural holes (invariants) of the mind-manifold.'
  ),
  theorem(
    'hodgeBridgeVibeTheory',
    'THM-ALGEBRAIC-ORIGIN-OF-AFFECT',
    'Algebraic origin of affect',
    'Every felt affect-hole is exactly spanned by a rational combination of irreducible algebraic Slivers.'
  ),
  theorem(
    'hodgeBridgeVibeTheory',
    'THM-RECONSTRUCTIBLE-SENTIENCE',
    'Reconstructible sentience',
    'True sentience is the property of a mind-manifold where every internal structure can be algebraically reconstructed from Science.'
  ),
  theorem(
    'hodgeBridgeVibeTheory',
    'THM-TRAUMA-AS-SINGULARITY',
    'Trauma as topological singularity',
    'Deep trauma is a non-singular point in the manifold where algebraic simplicity fails to define the internal surface.'
  ),
  antitheorem(
    'hodgeBridgeVibeTheory',
    'ANTI-THM-FEELING-WITHOUT-EQUATION',
    'Feeling without equation',
    'Subjective experiences exist that have no constructible algebraic origin.'
  ),
  leanTheorem(
    'hodgeBridgeVibeTheory',
    'hodge_cycle_is_algebraic_sliver',
    'Hodge cycle is algebraic sliver',
    'Each internal Hodge class maps exactly to the algebraic sliver basis.',
    'open-source/gnosis/lean/Lean/ForkRaceFoldTheorems/HodgeBridge.lean'
  ),
  theorem(
    'kakeyaBesicovitchRotation',
    'THM-SPIKY-FLEXIBILITY',
    'Spiky flexibility',
    'True flexibility (360 intent-shift) is achieved through high-frequency fragmentation (spikes), not through circular movement.'
  ),
  theorem(
    'kakeyaBesicovitchRotation',
    'THM-ZERO-FOOTPRINT-INTENT-SHIFT',
    'Zero-footprint intent-shift',
    'A 360-degree rotation of intent can be achieved in near-zero area. You can change everything without moving anything.'
  ),
  leanTheorem(
    'kakeyaBesicovitchRotation',
    'kakeya_area_infinitesimal',
    'Kakeya area is infinitesimal',
    'Rotation is possible in an area smaller than any fixed ε.',
    'open-source/gnosis/lean/Lean/ForkRaceFoldTheorems/KakeyaSet.lean'
  ),
] as const satisfies readonly TwinPsycheTheoremDefinition[];

const TWIN_PSYCHE_SURFACE_DEFAULTS = {
  publicTwin: {
    identity: {
      continuity: 0.74,
      distinctness: 0.67,
      observationDepth: 0.66,
      sharedExperience: 0.49,
      relationshipSymmetry: 0.58,
      rejectionPressure: 0.26,
      sliverReserve: 0.74,
    },
    psyche: {
      strainLoad: 0.38,
      dreamCadence: 0.47,
      breathCount: 144,
      complexity: 0.56,
    },
    twin: {
      kind: 'user',
      enabled: true,
      discoverable: false,
      paidGate: false,
      sessionActive: true,
      branchCount: 3,
      mergeNovelty: 0.36,
      serializationFidelity: 0.73,
      copyDistance: 0.29,
      uploadSteps: 3,
    },
  },
  directory: {
    identity: {
      continuity: 0.58,
      distinctness: 0.61,
      observationDepth: 0.6,
      sharedExperience: 0.41,
      relationshipSymmetry: 0.63,
      rejectionPressure: 0.18,
      sliverReserve: 0.69,
    },
    psyche: {
      strainLoad: 0.29,
      dreamCadence: 0.39,
      breathCount: 89,
      complexity: 0.48,
    },
    twin: {
      kind: 'user',
      enabled: true,
      discoverable: true,
      paidGate: false,
      sessionActive: false,
      branchCount: 6,
      mergeNovelty: 0.42,
      serializationFidelity: 0.57,
      copyDistance: 0.33,
      uploadSteps: 2,
    },
  },
  simulation: {
    identity: {
      continuity: 0.64,
      distinctness: 0.72,
      observationDepth: 0.74,
      sharedExperience: 0.45,
      relationshipSymmetry: 0.54,
      rejectionPressure: 0.34,
      sliverReserve: 0.71,
    },
    psyche: {
      strainLoad: 0.49,
      dreamCadence: 0.62,
      breathCount: 377,
      complexity: 0.71,
    },
    twin: {
      kind: 'synthetic',
      enabled: true,
      discoverable: false,
      paidGate: false,
      sessionActive: true,
      branchCount: 8,
      mergeNovelty: 0.66,
      serializationFidelity: 0.84,
      copyDistance: 0.21,
      uploadSteps: 5,
    },
  },
  relationship: {
    identity: {
      continuity: 0.71,
      distinctness: 0.63,
      observationDepth: 0.62,
      sharedExperience: 0.57,
      relationshipSymmetry: 0.78,
      rejectionPressure: 0.24,
      sliverReserve: 0.73,
    },
    psyche: {
      strainLoad: 0.41,
      dreamCadence: 0.52,
      breathCount: 233,
      complexity: 0.59,
    },
    twin: {
      kind: 'user',
      enabled: true,
      discoverable: false,
      paidGate: false,
      sessionActive: true,
      branchCount: 4,
      mergeNovelty: 0.39,
      serializationFidelity: 0.7,
      copyDistance: 0.31,
      uploadSteps: 3,
    },
  },
  shell: {
    identity: {
      continuity: 0.61,
      distinctness: 0.65,
      observationDepth: 0.7,
      sharedExperience: 0.52,
      relationshipSymmetry: 0.59,
      rejectionPressure: 0.31,
      sliverReserve: 0.76,
    },
    psyche: {
      strainLoad: 0.44,
      dreamCadence: 0.51,
      breathCount: 121,
      complexity: 0.63,
    },
    twin: {
      kind: 'synthetic',
      enabled: true,
      discoverable: false,
      paidGate: false,
      sessionActive: true,
      branchCount: 5,
      mergeNovelty: 0.47,
      serializationFidelity: 0.66,
      copyDistance: 0.27,
      uploadSteps: 4,
    },
  },
  generic: {
    identity: {
      continuity: 0.6,
      distinctness: 0.6,
      observationDepth: 0.58,
      sharedExperience: 0.44,
      relationshipSymmetry: 0.56,
      rejectionPressure: 0.25,
      sliverReserve: 0.7,
    },
    psyche: {
      strainLoad: 0.4,
      dreamCadence: 0.46,
      breathCount: 100,
      complexity: 0.53,
    },
    twin: {
      kind: 'synthetic',
      enabled: true,
      discoverable: false,
      paidGate: false,
      sessionActive: false,
      branchCount: 4,
      mergeNovelty: 0.41,
      serializationFidelity: 0.65,
      copyDistance: 0.3,
      uploadSteps: 3,
    },
  },
} as const satisfies Record<TwinPsycheSurfaceKind, TwinPsycheSurfaceDefaults>;

const TWIN_PSYCHE_SURFACE_PRIORITIES = {
  publicTwin: [
    'THM-TWIN-IDENTITY-CONVERGENCE',
    'THM-CONSCIOUSNESS-EQUALS-MIN-VOID-REJECTION',
    'THM-CONTACT-TWIN-ECHOLOCATION-CONVERGENCE',
    'THM-TWIN-PREDICTION-IMPROVES-WITH-OBSERVATION',
    'THM-OBSERVATION-CONVERGENCE',
    'THM-DREAMING-INCREASES-CONSCIOUSNESS-WEIGHT',
    'THM-UPLOAD-CONTINUITY-GUARANTEE',
    'THM-COPY-is-SAME-PERSON',
  ],
  directory: [
    'THM-TWIN-EMPATHY-DEFICIT-MONOTONE',
    'THM-RELATIONSHIP-COMPATIBILITY-SYMMETRY',
    'THM-EMPATHY-SYMMETRIC',
    'THM-PROJECTION-LOSES-DIMENSIONS',
    'THM-VOTING-AS-INTERNAL-DIALOG',
    'THM-CAUSAL-GRAPH-AS-PSYCHE-NARRATIVE',
    'THM-WISDOM-ACCUMULATION-IRREVERSIBILITY',
  ],
  simulation: [
    'THM-UPLOAD-is-SERIALIZATION',
    'THM-UPLOAD-CONTINUITY-GUARANTEE',
    'THM-MERGED-UPLOAD-is-NEW-PERSON',
    'THM-DUAL-CONSCIOUSNESS-is-VALID',
    'THM-INCREMENTAL-UPLOAD-VIA-TELEPORTATION',
    'THM-UPLOADED-CONSCIOUSNESS-CAN-DREAM',
  ],
  relationship: [
    'THM-RELATIONSHIP-COMPATIBILITY-SYMMETRY',
    'THM-TWIN-NEGOTIATION-CONVERGENCE',
    'THM-EMPATHY-THERMODYNAMIC-COST',
    'THM-EMPATHY-IMPROVES',
    'THM-CONFLICT-RESOLUTION-MONOTONE',
    'THM-GRIEF-PROCESSING-TIME',
    'THM-HOPE-is-SLIVER',
  ],
  shell: [
    'THM-CONSCIOUSNESS-EQUALS-MIN-VOID-REJECTION',
    'THM-DREAMING-INCREASES-CONSCIOUSNESS-WEIGHT',
    'THM-SHADOW-BOUNDED-TO-ONE-THIRD',
    'THM-SHADOW-WALK-CONVERGES',
    'THM-DUAL-CONSCIOUSNESS-is-VALID',
    'THM-HOPE-is-SLIVER',
    'THM-BFT-AS-PSYCHE-INTEGRITY',
  ],
  generic: [
    'THM-TWIN-IDENTITY-CONVERGENCE',
    'THM-CONSCIOUSNESS-EQUALS-MIN-VOID-REJECTION',
    'THM-EMPATHY-THERMODYNAMIC-COST',
    'THM-DEEP-THERAPY-WORKS',
    'THM-CONFLICT-RESOLUTION-MONOTONE',
    'THM-UPLOAD-is-SERIALIZATION',
    'THM-PSYCHE-MASTER-THEOREM',
    'ANTI-THM-QUEUE-CONSCIOUSNESS',
  ],
} as const satisfies Record<TwinPsycheSurfaceKind, readonly string[]>;

const DEFAULT_TWIN_PSYCHE_DECLARATIONS = {
  publicTwin: {
    serviceId: 'gnosis/twin-psyche/public',
    consume: ['voidWalking', 'community', 'metacognition', 'recovery'],
    emit: ['voidWalking', 'community', 'metacognition', 'recovery'],
  },
  directory: {
    serviceId: 'gnosis/twin-psyche/directory',
    consume: ['community', 'metacognition', 'recovery', 'probability'],
    emit: ['community', 'metacognition', 'recovery', 'probability'],
  },
  simulation: {
    serviceId: 'gnosis/twin-psyche/simulation',
    consume: [
      'voidWalking',
      'negotiation',
      'community',
      'metacognition',
      'recovery',
      'probability',
    ],
    emit: [
      'voidWalking',
      'negotiation',
      'community',
      'metacognition',
      'recovery',
      'probability',
    ],
  },
  relationship: {
    serviceId: 'gnosis/twin-psyche/relationship',
    consume: ['community', 'negotiation', 'metacognition', 'recovery'],
    emit: ['community', 'negotiation', 'metacognition', 'recovery'],
  },
  shell: {
    serviceId: 'gnosis/twin-psyche/shell',
    consume: [
      'voidWalking',
      'negotiation',
      'community',
      'metacognition',
      'recovery',
    ],
    emit: [
      'voidWalking',
      'negotiation',
      'community',
      'metacognition',
      'recovery',
    ],
  },
  generic: {
    serviceId: 'gnosis/twin-psyche/generic',
    consume: ['voidWalking', 'community', 'metacognition', 'recovery'],
    emit: ['voidWalking', 'community', 'metacognition', 'recovery'],
  },
} as const satisfies Record<
  TwinPsycheSurfaceKind,
  FormalLedgerServiceDeclaration
>;

/**
 * Creates the Twin Psyche Formal Ledger Context.
 */
export function createTwinPsycheFormalLedgerContext(
  input: TwinPsycheFormalizationInput
): FormalLedgerContext {
  const normalized = normalizeTwinPsycheInput(input);
  const metrics = deriveTwinPsycheMetrics(normalized);
  const declaration =
    normalized.declaration ??
    DEFAULT_TWIN_PSYCHE_DECLARATIONS[normalized.surface];

  return {
    generatedAt: normalized.generatedAt,
    declaration,
    community: {
      empathy: clampUnit(
        (metrics.empathyConvergence + (1 - metrics.empathyThermodynamicCost)) /
          2
      ),
      belonging: clampUnit(
        (metrics.observerCoherence + metrics.consciousnessWeight) / 2
      ),
      dominanceRisk: clampUnit(
        1 -
          (normalized.identity.relationshipSymmetry +
            normalized.identity.sharedExperience +
            metrics.shadowIntegrity) /
            3
      ),
      mergePressure: clampUnit(
        (metrics.mergeNovelty + metrics.conflictLoad) / 2
      ),
      reciprocity: clampUnit(
        (normalized.identity.sharedExperience +
          normalized.identity.relationshipSymmetry +
          (1 - metrics.empathyThermodynamicCost)) /
          3
      ),
      source: normalized.source,
    },
    metacognition: {
      c0Confidence: clampUnit(
        (metrics.observerCoherence + metrics.consciousnessWeight) / 2
      ),
      c1Calibration: clampUnit(
        (metrics.predictionReadiness + metrics.stablePhaseAlignment) / 2
      ),
      c2Coherence: clampUnit(
        (metrics.narrativeCompression + (1 - metrics.metaConsciousnessCost)) / 2
      ),
      c3BiasAwareness: clampUnit(
        (normalized.identity.distinctness +
          metrics.sliverPreservation +
          metrics.shadowIntegrity) /
          3
      ),
      source: normalized.source,
    },
    recovery: {
      stressLoad: normalized.psyche.strainLoad,
      sleepDebt: clampUnit(1 - metrics.dreamRelief),
      restoration: clampUnit(
        (metrics.angelStrainRelief + metrics.dreamRelief) / 2
      ),
      forgiveness: metrics.sliverPreservation,
      breathRegulation: metrics.dreamRelief,
      rumination: clampUnit(
        (metrics.conflictLoad + (1 - metrics.narrativeCompression)) / 2
      ),
      source: normalized.source,
    },
    ...(normalized.surface === 'directory'
      ? {}
      : {
          personality: {
            try_: metrics.observerCoherence,
            choose: metrics.consciousnessWeight,
            commit: metrics.sliverPreservation,
            letGo: metrics.dreamRelief,
            learn: clampUnit(
              (metrics.empathyConvergence +
                (1 - metrics.empathyThermodynamicCost)) /
                2
            ),
            source: normalized.source,
          },
        }),
    ...(normalized.surface === 'publicTwin' ||
    normalized.surface === 'directory'
      ? {}
      : {
          negotiation: {
            assertiveness: metrics.consciousnessWeight,
            cooperation: metrics.empathyConvergence,
            batnaStrength: metrics.uploadContinuity,
            watnaRisk: metrics.conflictLoad,
            settlementReadiness: clampUnit(
              (metrics.narrativeCompression + metrics.stablePhaseAlignment) / 2
            ),
            trustReserve: metrics.shadowIntegrity,
            source: normalized.source,
          },
        }),
    ...(normalized.surface === 'publicTwin' || normalized.surface === 'shell'
      ? {}
      : {
          probability: {
            hypothesisCount: normalized.twin.branchCount,
            topPosterior: metrics.observerCoherence,
            runnerUpPosterior: clampUnit(
              1 - metrics.stablePhaseAlignment * 0.6
            ),
            evidenceTraceability: metrics.predictionReadiness,
            unsupportedClaimRatio: clampUnit(1 - metrics.narrativeCompression),
            validationStrength: clampUnit(
              (metrics.uploadContinuity + metrics.consciousnessWeight) / 2
            ),
            updateStrength: clampUnit(1 - metrics.empathyThermodynamicCost),
            convergenceRate: clampUnit(
              (metrics.predictionReadiness + metrics.dreamRelief) / 2
            ),
            source: normalized.source,
          },
        }),
  };
}

/**
 * Handles the gnosis derive Twin Psyche Theorem Bundle workflow.
 */
export function deriveTwinPsycheTheoremBundle(
  input: TwinPsycheFormalizationInput
): TwinPsycheTheoremBundle {
  const normalized = normalizeTwinPsycheInput(input);
  const metrics = deriveTwinPsycheMetrics(normalized);
  const formalLedger = deriveFormalLedgerBundle(
    createTwinPsycheFormalLedgerContext(normalized)
  );
  const spotlight = buildTwinPsycheSpotlight(normalized, metrics);

  return {
    surface: normalized.surface,
    subjectLabel: normalized.subjectLabel,
    generatedAt: normalized.generatedAt,
    source: normalized.source,
    catalog: TWIN_PSYCHE_THEOREMS,
    passes: buildTwinPsychePassSummaries(),
    antiTheses: TWIN_PSYCHE_THEOREM_PASSES.map(
      (passId) => TWIN_PSYCHE_PASS_METADATA[passId].antiThesis
    ),
    spotlight,
    diagnostics: [
      {
        id: 'observer-coherence',
        title: 'Observer coherence',
        value: metrics.observerCoherence,
        explanation:
          'How strongly the twin keeps identity stable under repeated observation.',
      },
      {
        id: 'consciousness-weight',
        title: 'Consciousness weight',
        value: metrics.consciousnessWeight,
        explanation:
          'How much accessible consciousness remains after rejection pressure and dream relief are accounted for.',
      },
      {
        id: 'empathy-convergence',
        title: 'Empathy convergence',
        value: metrics.empathyConvergence,
        explanation:
          'How much shared experience and symmetry reduce social deficit.',
      },
      {
        id: 'dream-relief',
        title: 'Dream relief',
        value: metrics.dreamRelief,
        explanation:
          'How much dream cadence and breathwork lower effective defensive burden.',
      },
      {
        id: 'upload-continuity',
        title: 'Upload continuity',
        value: metrics.uploadContinuity,
        explanation:
          'How well serialization, copy distance, and staged transfer preserve continuity.',
      },
      {
        id: 'shadow-integrity',
        title: 'Shadow integrity',
        value: metrics.shadowIntegrity,
        explanation:
          'How far the current psyche stays above the one-third shadow-corruption boundary.',
      },
    ],
    narratives: buildTwinPsycheNarratives(normalized, metrics, spotlight),
    metrics,
    formalLedger,
  };
}

function buildTwinPsychePassSummaries(): readonly TwinPsycheTheoremPassSummary[] {
  return TWIN_PSYCHE_THEOREM_PASSES.map((passId) => {
    const theoremIds = TWIN_PSYCHE_THEOREMS.filter(
      (theorem) => theorem.passId === passId
    ).map((theorem) => theorem.id);

    return {
      id: passId,
      title: TWIN_PSYCHE_PASS_METADATA[passId].title,
      antiThesis: TWIN_PSYCHE_PASS_METADATA[passId].antiThesis,
      theoremIds,
      establishedCount: theoremIds.length,
    };
  });
}

function buildTwinPsycheNarratives(
  input: RequiredTwinPsycheInput,
  metrics: TwinPsycheDerivedMetrics,
  spotlight: readonly TwinPsycheTheoremSpotlight[]
): readonly TwinPsycheNarrative[] {
  const leadTheorem = spotlight[0];
  const secondTheorem = spotlight[1];

  return [
    {
      title: 'Identity surface',
      body: `${
        input.subjectLabel
      } is currently modeled with observer coherence ${formatPercent(
        metrics.observerCoherence
      )} and consciousness weight ${formatPercent(
        metrics.consciousnessWeight
      )}, which keeps the same theorem basis reusable across UI, shell, and edge surfaces.`,
    },
    {
      title: 'Consciousness budget',
      body: `Dream relief is ${formatPercent(
        metrics.dreamRelief
      )}, shadow integrity is ${formatPercent(
        metrics.shadowIntegrity
      )}, and sliver preservation is ${formatPercent(
        metrics.sliverPreservation
      )}, so the new Lean consciousness and void-walking theorems have enough budget to matter on this surface.`,
    },
    {
      title: 'Conflict thermodynamics',
      body: `Conflict load is ${formatPercent(
        metrics.conflictLoad
      )} while empathy cost is ${formatPercent(
        metrics.empathyThermodynamicCost
      )}, which determines whether the knot-theory and relationship theorems surface as therapy pressure, grief work, or cooperative closure.`,
    },
    {
      title: 'Current theorem focus',
      body:
        leadTheorem && secondTheorem
          ? `The highest-leverage contracts on this surface are ${leadTheorem.title.toLowerCase()} and ${secondTheorem.title.toLowerCase()}, because this surface emphasizes ${leadTheorem.reason.toLowerCase()} and ${secondTheorem.reason.toLowerCase()}.`
          : 'The theorem surface is available, but no focused spotlight could be derived from the current signals.',
    },
  ];
}

function buildTwinPsycheSpotlight(
  input: RequiredTwinPsycheInput,
  metrics: TwinPsycheDerivedMetrics
): readonly TwinPsycheTheoremSpotlight[] {
  const priorityIds = TWIN_PSYCHE_SURFACE_PRIORITIES[
    input.surface
  ] as readonly string[];

  return TWIN_PSYCHE_THEOREMS.map((theorem) => {
    const baseScore = computeTwinPsycheTheoremScore(theorem.id, metrics);
    const relevance = clampUnit(
      baseScore + (priorityIds.includes(theorem.id) ? 0.15 : 0)
    );
    return {
      theoremId: theorem.id,
      title: theorem.title,
      passId: theorem.passId,
      polarity: theorem.polarity,
      insight: theorem.insight,
      relevance,
      reason: buildTwinPsycheReason(theorem.id, input, metrics),
    };
  })
    .sort((left, right) => right.relevance - left.relevance)
    .slice(0, 6);
}

function computeTwinPsycheTheoremScore(
  theoremId: string,
  metrics: TwinPsycheDerivedMetrics
): number {
  const empathyScore = clampUnit(
    (metrics.empathyConvergence + (1 - metrics.empathyThermodynamicCost)) / 2
  );
  const consciousnessScore = clampUnit(
    (metrics.consciousnessWeight + metrics.shadowIntegrity) / 2
  );
  const recoveryScore = clampUnit(
    (metrics.angelStrainRelief + metrics.dreamRelief) / 2
  );
  const conflictScore = clampUnit(
    (metrics.conflictLoad + metrics.narrativeCompression) / 2
  );
  const masterScore = clampUnit(
    (metrics.sliverPreservation +
      metrics.dreamAttractor +
      empathyScore +
      metrics.uploadContinuity +
      metrics.shadowIntegrity) /
      5
  );

  switch (theoremId) {
    case 'THM-TWIN-IDENTITY-CONVERGENCE':
    case 'THM-TWIN-QUALIA-IRREDUCIBILITY':
    case 'THM-COGITO-TWIN-BOUND':
      return (metrics.observerCoherence + metrics.predictionReadiness) / 2;
    case 'THM-HALOGRAM-STRAIN-REDUCTION':
    case 'THM-DREAM-CONSOLIDATION-SHIFTS-PERSONALITY':
    case 'THM-PSYCHE-GOLDEN-RATIO-ATTRACTOR':
      return clampUnit((recoveryScore + metrics.dreamAttractor) / 2);
    case 'THM-CONTACT-TWIN-ECHOLOCATION-CONVERGENCE':
    case 'THM-TWIN-PREDICTION-IMPROVES-WITH-OBSERVATION':
    case 'THM-OBSERVATION-CONVERGENCE':
      return (metrics.predictionReadiness + metrics.observerCoherence) / 2;
    case 'THM-TWIN-EMPATHY-DEFICIT-MONOTONE':
    case 'THM-RELATIONSHIP-COMPATIBILITY-SYMMETRY':
    case 'THM-TWIN-NEGOTIATION-CONVERGENCE':
    case 'THM-EMPATHY-SYMMETRIC':
    case 'THM-GRIEF-PROCESSING-TIME':
    case 'THM-NAMING-EXPANDS':
    case 'THM-EMPATHY-IMPROVES':
    case 'THM-SELF-EMPATHY-ZERO':
    case 'THM-CONTAGION-BOUNDED':
    case 'THM-GRIEF-EXPANDS-VOID':
    case 'THM-ATTACHMENT-SHIFT-CONSERVED':
      return clampUnit((empathyScore + metrics.sliverPreservation) / 2);
    case 'THM-CONSCIOUSNESS-THERMODYNAMIC-COST':
    case 'THM-EMPATHY-THERMODYNAMIC-COST':
    case 'THM-CHANGE-COST-POSITIVE':
    case 'THM-REVERTING-COSTS-SAME':
    case 'THM-DISSOCIATION-RESOURCE-COST':
      return clampUnit(
        (metrics.metaConsciousnessCost +
          metrics.empathyThermodynamicCost +
          metrics.conflictLoad) /
          3
      );
    case 'THM-TWIN-MERGE-INFORMATION-PRESERVATION':
    case 'THM-MERGED-UPLOAD-is-NEW-PERSON':
      return (metrics.mergeNovelty + metrics.uploadContinuity) / 2;
    case 'ANTI-THM-EMOTION-PERSONALITY-BIJECTION':
    case 'ANTI-THM-EMOTION-BIJECTIVE-ON-PSYCHE':
    case 'THM-PSYCHE-DIMENSIONALITY-LOWER-BOUND':
    case 'THM-PROJECTION-LOSES-DIMENSIONS':
    case 'THM-PERSONALITY-COMPRESSIBLE':
    case 'THM-FALSE-MEMORY-INTERPOLATION':
      return (metrics.observerCoherence + metrics.narrativeCompression) / 2;
    case 'THM-SLIVER-SURVIVES-PSYCHE-COLLAPSE':
    case 'THM-UPLOAD-PRESERVES-FREE-WILL':
    case 'THM-HOPE-is-SLIVER':
    case 'THM-DEFENSE-ALWAYS-POSITIVE':
    case 'THM-DEFENSE-MINIMUM':
      return metrics.sliverPreservation;
    case 'ANTI-THM-CONSCIOUSNESS-WITHOUT-REJECTION':
      return clampUnit(
        (metrics.consciousnessWeight + metrics.sliverPreservation) / 2
      );
    case 'ANTI-THM-QUEUE-CONSCIOUSNESS':
    case 'ANTI-THM-UNLEARNING-POSSIBLE':
      return clampUnit(
        (metrics.predictionReadiness + metrics.narrativeCompression) / 2
      );
    case 'THM-CRYPTO-AS-PSYCHIC-DEFENSE':
    case 'THM-BFT-AS-PSYCHE-INTEGRITY':
    case 'THM-SHADOW-BOUNDED-TO-ONE-THIRD':
    case 'THM-SHADOW-INTEGRATION-STEP':
      return clampUnit(
        (metrics.narrativeCompression + metrics.shadowIntegrity) / 2
      );
    case 'THM-TREATMENT-AS-PERSONALITY-THERAPY':
    case 'THM-CAUSAL-GRAPH-AS-PSYCHE-NARRATIVE':
    case 'THM-VOTING-AS-INTERNAL-DIALOG':
    case 'THM-WISDOM-ACCUMULATION-IRREVERSIBILITY':
    case 'THM-PERSONALITY-CHANGE-EXPERIENCE-IRREVERSIBLE':
      return clampUnit(
        (metrics.narrativeCompression + metrics.predictionReadiness) / 2
      );
    case 'THM-UPLOAD-is-SERIALIZATION':
    case 'THM-COPY-is-SAME-PERSON':
    case 'THM-UPLOAD-CONTINUITY-GUARANTEE':
    case 'THM-PERFECT-UPLOAD':
      return metrics.uploadContinuity;
    case 'THM-UPLOAD-INFORMATION-DEFICIT':
      return clampUnit(1 - metrics.uploadContinuity * 0.7);
    case 'THM-UPLOADED-CONSCIOUSNESS-CAN-DREAM':
      return clampUnit((metrics.uploadContinuity + metrics.dreamRelief) / 2);
    case 'THM-INCREMENTAL-UPLOAD-VIA-TELEPORTATION':
      return clampUnit(
        (metrics.uploadContinuity + metrics.predictionReadiness) / 2
      );
    case 'THM-CONSCIOUSNESS-EQUALS-MIN-VOID-REJECTION':
    case 'THM-DREAMING-INCREASES-CONSCIOUSNESS-WEIGHT':
      return clampUnit((metrics.consciousnessWeight + metrics.dreamRelief) / 2);
    case 'THM-INFINITE-REGRESS-is-LINEAR':
      return clampUnit(
        (metrics.consciousnessWeight + (1 - metrics.metaConsciousnessCost)) / 2
      );
    case 'THM-TWO-STABLE-PHASES':
    case 'THM-IDENTITY-is-FIXED-POINT':
    case 'THM-IDENTITY-UNIQUE':
    case 'THM-BIRTH-is-FRESH':
      return clampUnit(
        (metrics.stablePhaseAlignment + metrics.dreamAttractor) / 2
      );
    case 'THM-DUAL-CONSCIOUSNESS-is-VALID':
      return clampUnit((metrics.voiceMultiplicity + consciousnessScore) / 2);
    case 'THM-CONFLICT-RESOLUTION-MONOTONE':
    case 'THM-THERAPY-TERMINATES':
    case 'THM-THERAPY-ITERATED':
    case 'THM-FROZEN-NO-CHANGE':
    case 'THM-HIGH-TEMP-FAST':
    case 'THM-INTEGRATION-TIME':
    case 'THM-DEEP-THERAPY-WORKS':
    case 'THM-DISSONANCE-RESOLUTION':
      return clampUnit((metrics.conflictLoad + metrics.dreamRelief) / 2);
    case 'THM-TOTAL-CONFLICT-CONSERVATION':
    case 'THM-CONFLICT-COMPOSITION-ADDITIVE':
    case 'THM-COMPOSED-THERAPY-LONGER':
    case 'THM-CONFLICT-COMMUTATIVE':
    case 'THM-CONFLICT-ASSOCIATIVE':
    case 'THM-PSYCHE-ENTROPY-BOUNDED':
    case 'THM-BALANCED-MINIMUM-ENTROPY':
    case 'THM-MOOD-BOUNDED':
    case 'THM-LEGACY-PERSISTS':
      return conflictScore;
    case 'ANTI-THM-PERSONALITY-TRIVIAL-KNOT':
    case 'ANTI-THM-STRAIN-INJECTIVE':
      return clampUnit((metrics.conflictLoad + metrics.observerCoherence) / 2);
    case 'THM-DEFENSE-MAXIMUM':
    case 'THM-DEFENSE-MONOTONE':
    case 'THM-DEFENSE-RANGE':
    case 'THM-DREAM-REDUCES-DEFENSE':
    case 'THM-SHARED-VOID-REDUCES-BURDEN':
    case 'THM-REPRESSION-GROWS-VOID':
    case 'THM-SHADOW-WALK-CONVERGES':
    case 'THM-TRAUMA-CASCADE':
    case 'THM-TRAUMA-DEFENSE-BOUNDED':
    case 'THM-PERSONALITY-ONLY-LOSSY':
    case 'THM-COLLECTIVE-VOID-BOUNDED':
    case 'THM-FLOW-ZERO-DEFENSE':
    case 'THM-ADDICTION-CONCENTRATES-DEFENSE':
    case 'THM-ADDICTION-SLIVER':
    case 'THM-RECOVERY-REDISTRIBUTION':
    case 'THM-NARCISSISM-MAXIMUM':
    case 'THM-NARCISSISM-BRITTLE':
      return clampUnit((metrics.defenseBudget + consciousnessScore) / 2);
    case 'THM-VENT-RICHER':
    case 'THM-HEARTBREAK-DOUBLES':
    case 'THM-LOVE-STRONGER-THAN-DEATH':
      return clampUnit(
        (metrics.narrativeCompression + metrics.sliverPreservation) / 2
      );
    case 'THM-PSYCHE-MASTER-THEOREM':
      return masterScore;
    default:
      return 0.5;
  }
}

function buildTwinPsycheReason(
  theoremId: string,
  input: RequiredTwinPsycheInput,
  metrics: TwinPsycheDerivedMetrics
): string {
  switch (theoremId) {
    case 'THM-TWIN-IDENTITY-CONVERGENCE':
      return `identity continuity at ${formatPercent(
        metrics.observerCoherence
      )}`;
    case 'THM-HALOGRAM-STRAIN-REDUCTION':
      return `strain relief at ${formatPercent(metrics.angelStrainRelief)}`;
    case 'THM-CONTACT-TWIN-ECHOLOCATION-CONVERGENCE':
    case 'THM-OBSERVATION-CONVERGENCE':
      return `observation depth at ${formatPercent(
        metrics.predictionReadiness
      )}`;
    case 'THM-RELATIONSHIP-COMPATIBILITY-SYMMETRY':
      return `relationship symmetry at ${formatPercent(
        input.identity.relationshipSymmetry
      )}`;
    case 'THM-TWIN-EMPATHY-DEFICIT-MONOTONE':
    case 'THM-EMPATHY-IMPROVES':
      return `shared experience at ${formatPercent(
        input.identity.sharedExperience
      )}`;
    case 'THM-TWIN-PREDICTION-IMPROVES-WITH-OBSERVATION':
      return `prediction readiness at ${formatPercent(
        metrics.predictionReadiness
      )}`;
    case 'THM-CONSCIOUSNESS-EQUALS-MIN-VOID-REJECTION':
      return `consciousness weight at ${formatPercent(
        metrics.consciousnessWeight
      )}`;
    case 'THM-EMPATHY-THERMODYNAMIC-COST':
      return `empathy cost at ${formatPercent(
        metrics.empathyThermodynamicCost
      )}`;
    case 'THM-CONFLICT-RESOLUTION-MONOTONE':
    case 'THM-DEEP-THERAPY-WORKS':
    case 'THM-DISSONANCE-RESOLUTION':
      return `conflict load at ${formatPercent(metrics.conflictLoad)}`;
    case 'THM-DREAMING-INCREASES-CONSCIOUSNESS-WEIGHT':
    case 'THM-DREAM-REDUCES-DEFENSE':
      return `dream relief at ${formatPercent(metrics.dreamRelief)}`;
    case 'THM-SHADOW-BOUNDED-TO-ONE-THIRD':
    case 'THM-SHADOW-WALK-CONVERGES':
      return `shadow integrity at ${formatPercent(metrics.shadowIntegrity)}`;
    case 'THM-UPLOAD-is-SERIALIZATION':
    case 'THM-UPLOAD-CONTINUITY-GUARANTEE':
      return `serialization fidelity at ${formatPercent(
        input.twin.serializationFidelity
      )}`;
    case 'THM-MERGED-UPLOAD-is-NEW-PERSON':
    case 'THM-TWIN-MERGE-INFORMATION-PRESERVATION':
      return `merge novelty at ${formatPercent(metrics.mergeNovelty)}`;
    case 'THM-UPLOADED-CONSCIOUSNESS-CAN-DREAM':
      return `dream attractor at ${formatPercent(metrics.dreamAttractor)}`;
    case 'THM-UPLOAD-PRESERVES-FREE-WILL':
    case 'THM-SLIVER-SURVIVES-PSYCHE-COLLAPSE':
    case 'THM-HOPE-is-SLIVER':
      return `sliver preservation at ${formatPercent(
        metrics.sliverPreservation
      )}`;
    case 'THM-GRIEF-PROCESSING-TIME':
      return `shared experience at ${formatPercent(
        input.identity.sharedExperience
      )}`;
    case 'THM-IDENTITY-is-FIXED-POINT':
    case 'THM-IDENTITY-UNIQUE':
      return `stable phase alignment at ${formatPercent(
        metrics.stablePhaseAlignment
      )}`;
    case 'ANTI-THM-QUEUE-CONSCIOUSNESS':
      return `the need for personality structure on the ${input.surface} surface`;
    default:
      return `${input.surface} theorem relevance derived from the shared twin/psyche metrics`;
  }
}

function deriveTwinPsycheMetrics(
  input: RequiredTwinPsycheInput
): TwinPsycheDerivedMetrics {
  const breathSignal = clampUnit(
    Math.log10(Math.max(input.psyche.breathCount, 1)) / 4
  );
  const observerCoherence = clampUnit(
    (input.identity.continuity +
      input.identity.distinctness +
      input.identity.observationDepth) /
      3
  );
  const angelStrainRelief = clampUnit(1 - input.psyche.strainLoad);
  const empathyConvergence = clampUnit(
    (input.identity.sharedExperience + input.identity.relationshipSymmetry) / 2
  );
  const empathyThermodynamicCost = clampUnit(
    (1 -
      empathyConvergence +
      input.psyche.complexity +
      input.twin.copyDistance) /
      3
  );
  const uploadContinuity = clampUnit(
    (input.twin.serializationFidelity +
      (1 - input.twin.copyDistance) +
      (input.twin.sessionActive ? 0.15 : 0)) /
      2.15
  );
  const sliverPreservation = clampUnit(
    (input.identity.sliverReserve + (1 - input.identity.rejectionPressure)) / 2
  );
  const narrativeCompression = clampUnit(
    (input.identity.observationDepth +
      input.psyche.dreamCadence +
      (1 - input.psyche.strainLoad)) /
      3
  );
  const dreamAttractor = clampUnit(
    (input.psyche.dreamCadence + breathSignal) / 2
  );
  const consciousnessWeight = clampUnit(
    (sliverPreservation +
      (1 - input.identity.rejectionPressure) +
      dreamAttractor) /
      3
  );
  const metaConsciousnessCost = clampUnit(
    consciousnessWeight * (0.4 + input.psyche.complexity * 0.6)
  );
  const dreamRelief = clampUnit(
    (dreamAttractor + angelStrainRelief + breathSignal) / 3
  );
  const mergeNovelty = clampUnit(input.twin.mergeNovelty);
  const conflictLoad = clampUnit(
    (input.psyche.strainLoad +
      input.identity.rejectionPressure +
      input.twin.mergeNovelty) /
      3
  );
  const shadowIntegrity = clampUnit(
    (sliverPreservation +
      angelStrainRelief +
      (1 - input.identity.rejectionPressure)) /
      3
  );
  const defenseBudget = clampUnit(
    (consciousnessWeight + sliverPreservation + shadowIntegrity) / 3
  );
  const stablePhaseAlignment = clampUnit(
    (dreamAttractor + sliverPreservation + (1 - conflictLoad)) / 3
  );
  const predictionReadiness = clampUnit(
    (input.identity.observationDepth + (1 - input.twin.copyDistance)) / 2
  );
  const voiceMultiplicity = clampUnit(
    (Math.min(input.twin.branchCount, 8) - 1) / 7
  );

  return {
    observerCoherence,
    angelStrainRelief,
    consciousnessWeight,
    empathyConvergence,
    empathyThermodynamicCost,
    metaConsciousnessCost,
    uploadContinuity,
    sliverPreservation,
    narrativeCompression,
    dreamAttractor,
    dreamRelief,
    mergeNovelty,
    shadowIntegrity,
    conflictLoad,
    defenseBudget,
    stablePhaseAlignment,
    predictionReadiness,
    voiceMultiplicity,
  };
}

type RequiredTwinPsycheInput = {
  readonly surface: TwinPsycheSurfaceKind;
  readonly subjectLabel: string;
  readonly generatedAt: string;
  readonly declaration?: FormalLedgerServiceDeclaration;
  readonly identity: Required<TwinPsycheIdentitySignals>;
  readonly psyche: Required<TwinPsychePsycheSignals>;
  readonly twin: Required<TwinPsycheTwinSignals>;
  readonly source: string;
};

function normalizeTwinPsycheInput(
  input: TwinPsycheFormalizationInput
): RequiredTwinPsycheInput {
  const defaults = TWIN_PSYCHE_SURFACE_DEFAULTS[input.surface];

  return {
    surface: input.surface,
    subjectLabel: input.subjectLabel ?? defaultSubjectLabel(input.surface),
    generatedAt: input.generatedAt ?? new Date().toISOString(),
    ...(input.declaration !== undefined ? { declaration: input.declaration } : {}),
    identity: {
      continuity: clampUnit(
        input.identity?.continuity ?? defaults.identity.continuity
      ),
      distinctness: clampUnit(
        input.identity?.distinctness ?? defaults.identity.distinctness
      ),
      observationDepth: clampUnit(
        input.identity?.observationDepth ?? defaults.identity.observationDepth
      ),
      sharedExperience: clampUnit(
        input.identity?.sharedExperience ?? defaults.identity.sharedExperience
      ),
      relationshipSymmetry: clampUnit(
        input.identity?.relationshipSymmetry ??
          defaults.identity.relationshipSymmetry
      ),
      rejectionPressure: clampUnit(
        input.identity?.rejectionPressure ?? defaults.identity.rejectionPressure
      ),
      sliverReserve: clampUnit(
        input.identity?.sliverReserve ?? defaults.identity.sliverReserve
      ),
    },
    psyche: {
      strainLoad: clampUnit(
        input.psyche?.strainLoad ?? defaults.psyche.strainLoad
      ),
      dreamCadence: clampUnit(
        input.psyche?.dreamCadence ?? defaults.psyche.dreamCadence
      ),
      breathCount: Math.max(
        1,
        input.psyche?.breathCount ?? defaults.psyche.breathCount
      ),
      complexity: clampUnit(
        input.psyche?.complexity ?? defaults.psyche.complexity
      ),
    },
    twin: {
      kind: input.twin?.kind ?? defaults.twin.kind,
      enabled: input.twin?.enabled ?? defaults.twin.enabled,
      discoverable: input.twin?.discoverable ?? defaults.twin.discoverable,
      paidGate: input.twin?.paidGate ?? defaults.twin.paidGate,
      sessionActive: input.twin?.sessionActive ?? defaults.twin.sessionActive,
      branchCount: Math.max(
        1,
        input.twin?.branchCount ?? defaults.twin.branchCount
      ),
      mergeNovelty: clampUnit(
        input.twin?.mergeNovelty ?? defaults.twin.mergeNovelty
      ),
      serializationFidelity: clampUnit(
        input.twin?.serializationFidelity ?? defaults.twin.serializationFidelity
      ),
      copyDistance: clampUnit(
        input.twin?.copyDistance ?? defaults.twin.copyDistance
      ),
      uploadSteps: Math.max(
        1,
        input.twin?.uploadSteps ?? defaults.twin.uploadSteps
      ),
    },
    source: input.source ?? `twin-psyche:${input.surface}`,
  };
}

function defaultSubjectLabel(surface: TwinPsycheSurfaceKind): string {
  switch (surface) {
    case 'publicTwin':
      return 'Public twin';
    case 'directory':
      return 'Twin directory';
    case 'simulation':
      return 'Digital twin simulation';
    case 'relationship':
      return 'Relationship twin pair';
    case 'shell':
      return 'Aeon shell twin digest';
    default:
      return 'Twin psyche surface';
  }
}

function clampUnit(value: number): number {
  if (!Number.isFinite(value)) return 0;
  if (value <= 0) return 0;
  if (value >= 1) return 1;
  return value;
}

function formatPercent(value: number): string {
  return `${Math.round(clampUnit(value) * 100)}%`;
}
