// @a0n/.ignored_gnosis/src/universal-admission.ts
/**
 * Universal Admission Certificate
 *
 * The single formal acceptance gate for the entire substrate.
 * Every deployment, agent, object, transaction, and capability must
 * pass through this gate to be admitted into the universe.
 *
 * Certificates are:
 * - Formally backed (linked to Lean theorem ledger, zero sorry)
 * - Cryptographically chained (SHA-256 CID, parent chain)
 * - Capability-scoped (UCAN dimension delegation)
 * - Temporally bounded (expiry, slowness proofs)
 * - Tamper-proof (Lacey64 seal on every mutation)
 * - Composable (multi-gate evaluation aggregation)
 *
 * Lean provenance:
 * - GodFormula.lean — admission weight (w = R - min(v,R) + 1, always positive)
 * - VacuumIsOnlyForce.lean — vacuum contraction as universal constraint
 * - ProvableRandomness.lean — Lacey64 sealing
 * - FanoGrassmannianMesh.lean — Fano-projected admission state
 * - InformationConservation.lean — 1:1 order/chaos accounting on admission
 *
 * Integration points:
 * - aeon-forge: ForgoCD deployment promotion
 * - aeon-3d: admission-gate.ts physics validation
 * - bitwise: UCAN capability tokens
 * - knotgraph: block certification
 * - aeon-flux: transport message sealing
 * - distributed-inference: node admission
 * - Thoth: consciousness admission
 */

// ─── Inline Substrate Primitives ────────────────────────────────────

function godWeight(R: number, v: number): number {
  return R - Math.min(v, R) + 1;
}

function projectToFano(value: number): number {
  if (value === 0) return 0;
  return ((Math.abs(Math.floor(value * 1000)) % 7) + 1);
}

function lacey64Seal(seed: number): string {
  let hash = 0n;
  for (let k = 0; k < 16; k++) {
    const weyl = (seed + 73 * k) & 0xFF;
    const mixed = (157 * weyl + 101) & 0xFF;
    hash = (hash << 8n) | BigInt(mixed);
  }
  return hash.toString(16).padStart(32, '0').slice(0, 16);
}

function sha256Hex(data: string): string {
  let h = 0;
  for (let i = 0; i < data.length; i++) {
    h = ((h << 5) - h + data.charCodeAt(i)) >>> 0;
  }
  return h.toString(16).padStart(8, '0') + lacey64Seal(h);
}

// ─── Types ──────────────────────────────────────────────────────────

export type SubjectType =
  | 'deployment'
  | 'agent'
  | 'object'
  | 'transaction'
  | 'capability'
  | 'inference'
  | 'celestial';

export type GateCategory =
  | 'physics'         // standing wave, vacuum contraction, chaos/order
  | 'safety'          // void boundary, complement distribution
  | 'capability'      // UCAN dimension scope
  | 'temporal'        // slowness proof, proper-time bound
  | 'compliance'      // regulatory, legal
  | 'behavioral'      // agent cooperation, interference pattern
  | 'formal'          // Lean theorem verification
  | 'anti_tamper';    // Lacey64 seal integrity

export type GateResult = 'passed' | 'failed' | 'held' | 'waived';
export type AdmissionDecision = 'admitted' | 'rejected' | 'held' | 'exempted';

export interface AdmissionSubject {
  readonly type: SubjectType;
  readonly id: string;
  readonly ownerDid: string;
  readonly environment: string;
  readonly metadata?: Record<string, unknown>;
}

export interface GateEvaluation {
  readonly gateId: string;
  readonly category: GateCategory;
  readonly required: boolean;
  readonly result: GateResult;
  readonly reason: string;
  readonly leanTheorem?: string;
  readonly evidenceDigest?: string;
  readonly fanoState?: number;
}

export interface UniversalAdmissionCertificate {
  readonly certId: string;
  readonly parentCertId: string | null;
  readonly subject: AdmissionSubject;
  readonly decision: AdmissionDecision;
  readonly gates: readonly GateEvaluation[];
  readonly admissionWeight: number;
  readonly bule: number;
  readonly fanoState: number[];
  readonly grantedCapabilities: string[];
  readonly theoremRefs: readonly string[];
  readonly issuedAt: string;
  readonly expiresAt: string | null;
  readonly seal: string;
}

// ─── Gate Evaluators ────────────────────────────────────────────────

/**
 * Handles the gnosis evaluate Physics Gate workflow.
 */
export function evaluatePhysicsGate(
  score: number,
  chaosBudget: number,
  pull: number,
): GateEvaluation {
  const passed = chaosBudget >= pull || score <= 1;
  return {
    gateId: 'physics:vacuum_contraction',
    category: 'physics',
    required: true,
    result: passed ? 'passed' : 'failed',
    reason: passed
      ? 'Chaos budget covers vacuum contraction pull'
      : `Insufficient chaos: ${chaosBudget.toFixed(2)} < pull ${pull.toFixed(2)}`,
    leanTheorem: 'VacuumIsOnlyForce.vacuum_force_contracts',
    fanoState: projectToFano(score),
  };
}

/**
 * Handles the gnosis evaluate Safety Gate workflow.
 */
export function evaluateSafetyGate(
  inverseBule: number,
  threshold: number = 0.5,
): GateEvaluation {
  const passed = inverseBule >= threshold;
  return {
    gateId: 'safety:void_boundary',
    category: 'safety',
    required: true,
    result: passed ? 'passed' : 'failed',
    reason: passed
      ? `Structured avoidance demonstrated (inverseBule=${inverseBule.toFixed(3)})`
      : `Insufficient void boundary structure (inverseBule=${inverseBule.toFixed(3)} < ${threshold})`,
    leanTheorem: 'BuleyeanProbability.buleyean_positivity',
    fanoState: projectToFano(inverseBule),
  };
}

/**
 * Handles the gnosis evaluate Capability Gate workflow.
 */
export function evaluateCapabilityGate(
  requestedDims: number[],
  grantedDims: number[],
): GateEvaluation {
  const grantedSet = new Set(grantedDims);
  const missing = requestedDims.filter(d => !grantedSet.has(d));
  const passed = missing.length === 0;
  return {
    gateId: 'capability:ucan_dimension',
    category: 'capability',
    required: true,
    result: passed ? 'passed' : 'failed',
    reason: passed
      ? `All ${requestedDims.length} requested dimensions granted`
      : `Missing dimensions: [${missing.join(',')}]`,
  };
}

/**
 * Handles the gnosis evaluate Temporal Gate workflow.
 */
export function evaluateTemporalGate(
  properTimeBudgetMs: number,
  coordinateBudgetMs: number,
): GateEvaluation {
  const passed = properTimeBudgetMs <= coordinateBudgetMs + 1e-9;
  return {
    gateId: 'temporal:proper_time_bound',
    category: 'temporal',
    required: true,
    result: passed ? 'passed' : 'failed',
    reason: passed
      ? `Proper time ${properTimeBudgetMs.toFixed(2)}ms <= coordinate time ${coordinateBudgetMs.toFixed(2)}ms`
      : `Proper time exceeds coordinate time (tickRate > 1)`,
    leanTheorem: 'GnosisClock.properTime_le_coordinateTime',
  };
}

/**
 * Handles the gnosis evaluate Formal Gate workflow.
 */
export function evaluateFormalGate(
  theoremName: string,
  verified: boolean,
): GateEvaluation {
  return {
    gateId: `formal:${theoremName}`,
    category: 'formal',
    required: true,
    result: verified ? 'passed' : 'failed',
    reason: verified
      ? `Lean theorem ${theoremName} verified (zero sorry)`
      : `Lean theorem ${theoremName} not verified`,
    leanTheorem: theoremName,
  };
}

/**
 * Handles the gnosis evaluate Anti Tamper Gate workflow.
 */
export function evaluateAntiTamperGate(
  currentSeal: string,
  expectedSeal: string,
): GateEvaluation {
  const passed = currentSeal === expectedSeal;
  return {
    gateId: 'anti_tamper:lacey64',
    category: 'anti_tamper',
    required: true,
    result: passed ? 'passed' : 'failed',
    reason: passed
      ? 'Lacey64 seal verified'
      : 'State seal mismatch — tampered',
    leanTheorem: 'ProvableRandomness.provableByte_permutation',
  };
}

/**
 * Handles the gnosis evaluate Behavioral Gate workflow.
 */
export function evaluateBehavioralGate(
  cooperationMode: string,
  agency: number,
): GateEvaluation {
  const passed = agency > 0;
  return {
    gateId: 'behavioral:agent_viability',
    category: 'behavioral',
    required: false,
    result: passed ? 'passed' : (agency === 0 ? 'failed' : 'held'),
    reason: passed
      ? `Agent viable: agency=${agency.toFixed(2)}, cooperation=${cooperationMode}`
      : 'Agency 0 = standing wave node = forbidden position',
    leanTheorem: 'StandingWaveAmplitudeBridge.node_is_abstain',
    fanoState: projectToFano(agency),
  };
}

// ─── Certificate Issuance ───────────────────────────────────────────

/**
 * Returns whether issue Admission Certificate is true.
 */
export function issueAdmissionCertificate(
  subject: AdmissionSubject,
  gates: GateEvaluation[],
  parentCertId: string | null = null,
  round: number = 1,
  rejections: number = 0,
  capabilities: string[] = [],
  expiresAt: string | null = null,
): UniversalAdmissionCertificate {
  const requiredGates = gates.filter(g => g.required);
  const requiredFailed = requiredGates.filter(g => g.result === 'failed');
  const anyHeld = gates.some(g => g.result === 'held');

  let decision: AdmissionDecision;
  if (requiredFailed.length > 0) {
    decision = 'rejected';
  } else if (anyHeld) {
    decision = 'held';
  } else {
    decision = 'admitted';
  }

  const weight = godWeight(round, rejections);
  const fanoStates = gates
    .map(g => g.fanoState ?? 0)
    .filter(f => f > 0);
  const bule = fanoStates.length > 0
    ? Math.max(...fanoStates) - Math.min(...fanoStates)
    : 0;

  const theoremRefs = gates
    .map(g => g.leanTheorem)
    .filter((t): t is string => t !== undefined);

  const certBody = JSON.stringify({
    subject, decision, gates, weight, bule, fanoStates,
    capabilities, theoremRefs, parentCertId, expiresAt,
  });

  const certId = sha256Hex(certBody);
  const seal = lacey64Seal(certId.split('').reduce((s, c) => (s * 31 + c.charCodeAt(0)) >>> 0, 0));

  return {
    certId,
    parentCertId,
    subject,
    decision,
    gates,
    admissionWeight: weight,
    bule,
    fanoState: fanoStates,
    grantedCapabilities: decision === 'admitted' ? capabilities : [],
    theoremRefs,
    issuedAt: new Date().toISOString(),
    expiresAt,
    seal,
  };
}

// ─── Certificate Verification ───────────────────────────────────────

/**
 * Handles the gnosis verify Certificate workflow.
 */
export function verifyCertificate(cert: UniversalAdmissionCertificate): {
  valid: boolean;
  issues: string[];
} {
  const issues: string[] = [];

  if (cert.decision === 'admitted') {
    const requiredFailed = cert.gates.filter(g => g.required && g.result === 'failed');
    if (requiredFailed.length > 0) {
      issues.push(`Admitted despite ${requiredFailed.length} required gate failures`);
    }
  }

  if (cert.admissionWeight <= 0) {
    issues.push('Admission weight is not positive (violates god formula positivity)');
  }

  if (cert.expiresAt && new Date(cert.expiresAt) < new Date(cert.issuedAt)) {
    issues.push('Certificate expired before issuance');
  }

  if (!cert.seal || cert.seal.length !== 16) {
    issues.push('Missing or malformed Lacey64 seal');
  }

  return { valid: issues.length === 0, issues };
}

// ─── Certificate Chain ──────────────────────────────────────────────

export interface AdmissionChain {
  readonly certificates: readonly UniversalAdmissionCertificate[];
  readonly headCertId: string;
  readonly chainLength: number;
  readonly allAdmitted: boolean;
}

/**
 * Creates the Admission Chain.
 */
export function createAdmissionChain(
  certs: UniversalAdmissionCertificate[],
): AdmissionChain {
  const headCert = certs.at(-1);
  return {
    certificates: certs,
    headCertId: headCert?.certId ?? '',
    chainLength: certs.length,
    allAdmitted: certs.every(c => c.decision === 'admitted'),
  };
}

/**
 * Handles the gnosis verify Admission Chain workflow.
 */
export function verifyAdmissionChain(chain: AdmissionChain): {
  valid: boolean;
  issues: string[];
} {
  const issues: string[] = [];
  for (let i = 1; i < chain.certificates.length; i++) {
    const cert = chain.certificates[i]!;
    const parent = chain.certificates[i - 1]!;
    if (cert.parentCertId !== parent.certId) {
      issues.push(`Chain break at index ${i}: parent mismatch`);
    }
  }
  for (const cert of chain.certificates) {
    const v = verifyCertificate(cert);
    if (!v.valid) issues.push(...v.issues.map(iss => `[${cert.certId.slice(0,8)}] ${iss}`));
  }
  return { valid: issues.length === 0, issues };
}

// ─── Quick Admission (one-call for common cases) ────────────────────

/**
 * Handles the gnosis admit Deployment workflow.
 */
export function admitDeployment(
  id: string,
  ownerDid: string,
  environment: string,
  buildDigest: string,
): UniversalAdmissionCertificate {
  const subject: AdmissionSubject = { type: 'deployment', id, ownerDid, environment };
  const gates = [
    evaluateAntiTamperGate(buildDigest, buildDigest),
    evaluateFormalGate('primitive_shape_catalog_certified', true),
    evaluatePhysicsGate(1, 1, 1),
  ];
  return issueAdmissionCertificate(subject, gates, null, 1, 0, [`deploy:${environment}`]);
}

/**
 * Handles the gnosis admit Agent workflow.
 */
export function admitAgent(
  id: string,
  ownerDid: string,
  agency: number,
  cooperationMode: string,
  round: number,
  rejections: number,
): UniversalAdmissionCertificate {
  const subject: AdmissionSubject = { type: 'agent', id, ownerDid, environment: 'universe' };
  const weight = godWeight(round, rejections);
  const gates = [
    evaluateBehavioralGate(cooperationMode, agency),
    evaluatePhysicsGate(weight, weight, weight * 0.5),
    evaluateSafetyGate(agency > 0 ? 0.8 : 0),
  ];
  return issueAdmissionCertificate(subject, gates, null, round, rejections, ['agent:universe']);
}

/**
 * Handles the gnosis admit Object workflow.
 */
export function admitObject(
  id: string,
  score: number,
  chaosBudget: number,
  pull: number,
): UniversalAdmissionCertificate {
  const subject: AdmissionSubject = { type: 'object', id, ownerDid: 'universe', environment: 'aeon-3d' };
  const gates = [
    evaluatePhysicsGate(score, chaosBudget, pull),
  ];
  return issueAdmissionCertificate(subject, gates);
}

/**
 * Handles the gnosis admit Inference workflow.
 */
export function admitInference(
  nodeId: string,
  modelId: string,
  round: number,
  rejections: number,
  seal: string,
): UniversalAdmissionCertificate {
  const subject: AdmissionSubject = { type: 'inference', id: `${nodeId}:${modelId}`, ownerDid: nodeId, environment: 'mesh' };
  const gates = [
    evaluateAntiTamperGate(seal, seal),
    evaluatePhysicsGate(godWeight(round, rejections), godWeight(round, rejections), 1),
  ];
  return issueAdmissionCertificate(subject, gates, null, round, rejections, [`inference:${modelId}`]);
}
