// @a0n/.ignored_gnosis/src/universal-admission.js
/**
 * Handles the gnosis evaluate Physics Gate workflow.
 */
export function evaluatePhysicsGate(score = 0, chaosBudget = 1, pull = 1) {
  const passed = Number.isFinite(score) && Number.isFinite(chaosBudget) && Number.isFinite(pull) && score >= 0 && chaosBudget >= 0 && pull >= 0;
  return {
    id: 'physics',
    passed,
    score,
    reason: passed ? 'physics gate accepted' : 'physics gate rejected',
  };
}

/**
 * Handles the gnosis evaluate Safety Gate workflow.
 */
export function evaluateSafetyGate(inverseBule = 0) {
  const passed = Number.isFinite(inverseBule) && inverseBule >= 0;
  return {
    id: 'safety',
    passed,
    score: inverseBule,
    reason: passed ? 'safety gate accepted' : 'safety gate rejected',
  };
}

/**
 * Handles the gnosis evaluate Anti Tamper Gate workflow.
 */
export function evaluateAntiTamperGate(expectedSeal, actualSeal) {
  const passed = expectedSeal === actualSeal;
  return {
    id: 'anti-tamper',
    passed,
    score: passed ? 1 : 0,
    reason: passed ? 'seal matched' : 'seal mismatch',
  };
}

/**
 * Returns whether issue Admission Certificate is true.
 */
export function issueAdmissionCertificate(subject, gates = []) {
  const admitted = gates.every((gate) => gate?.passed);
  return {
    subject,
    gates,
    decision: admitted ? 'admitted' : 'rejected',
    issuedAt: new Date().toISOString(),
  };
}
