// @a0n/.ignored_gnosis/src/universal-pisot-machine.ts
/**
 * universal-pisot-machine.ts -- Universal Pisot Machine (UPM) runtime.
 *
 * Formalized in Lean at Lean/ForkRaceFoldTheorems/UniversalPisotMachine.lean.
 *
 * A Turing-complete computing model based on the God Formula triad:
 *   FORK  (INC)     -- Increment a Pisot coordinate.
 *   RACE  (JUMP)    -- Fast-forward the Program Counter.
 *   FOLD  (SMOOTH)  -- Zeckendorf garbage collection / state transition.
 */

import { pisotF, pisotCheck, pisotCassini } from './betti/pisot-primitive';
import { PISOT_FIB } from './betti/pisot-primitive';

// ============================================================================
// UPM Instructions
// ============================================================================

export type UPMRegister = 'phi' | 'pell' | 'trib';

export type UPMInstruction =
  | { op: 'INC'; reg: UPMRegister }
  | { op: 'DEC_JUMP'; reg: UPMRegister; target: number }
  | { op: 'RACE'; k: number } // Logarithmic JUMP
  | { op: 'HALT' };

import { PulsePhase, evaluatePhase, type ClockPulse } from './quasicrystalline-clock';

// ================= ===========================================================
// UPM State
// ================= ===========================================================

export interface UPMState {
  phi: number;
  pell: number;
  trib: number;
  pc: number;
  halted: boolean;
  phase: PulsePhase;
}

/**
 * Handles the gnosis upm Init workflow.
 */
export function upmInit(): UPMState {
  return { phi: 0, pell: 0, trib: 0, pc: 0, halted: false, phase: PulsePhase.Even };
}

// ================= ===========================================================
// UPM Transition Function (Phase-Space Firewall protected)
// ================= ===========================================================

/**
 * Handles the gnosis upm Step workflow.
 */
export function upmStep(state: UPMState, program: UPMInstruction[]): UPMState {
  if (state.halted || state.pc >= program.length) {
    return { ...state, halted: true };
  }

  const instr = program[state.pc];
  const next = { ...state, pc: state.pc + 1 };

  // Phase-Space Firewall: Each UPM step must toggle the Cassini phase.
  // This turns replay attacks into "out-of-phase" rejection.
  next.phase = state.phase === PulsePhase.Even ? PulsePhase.Odd : PulsePhase.Even;

  // Cassini Parity Firewall: Verify the current register state before every step.
  pisotCheck(state.phi);

  switch (instr.op) {
    case 'INC':
      next[instr.reg] += 1;
      break;
    case 'DEC_JUMP':
      // Minsky Standard: if r > 0, r--. Else PC = target.
      if (state[instr.reg] > 0) {
        next[instr.reg] -= 1;
      } else {
        next.pc = instr.target;
      }
      break;
    case 'RACE':
      // Absolute JUMP (Warp Drive Leap).
      next.pc = instr.k;
      break;
    case 'HALT':
      next.halted = true;
      break;
  }

  // Verify the next state after the transition.
  pisotCheck(next.phi);

  return next;
}

/** Run a UPM program until completion or rejection. */
export function upmRun(program: UPMInstruction[], initialState = upmInit()): UPMState {
  let state = initialState;
  let steps = 0;
  const MAX_STEPS = 10000;

  while (!state.halted && steps < MAX_STEPS) {
    state = upmStep(state, program);
    steps++;
  }

  return state;
}

// ================= ===========================================================
// Universal Intelligence Wire Diet (Zeckendorf Representation)
// ================= ===========================================================

import { encodeZeckendorfArray, decodeZeckendorfArray } from '../../bitwise/src/bw-zeckendorf';

/**
 * Handles the gnosis encode UPMProgram workflow.
 */
export function encodeUPMProgram(program: UPMInstruction[]): Uint8Array {
  const payload: number[] = [];
  for (const instr of program) {
    switch (instr.op) {
      case 'INC':
        payload.push(instr.reg === 'phi' ? 0 : instr.reg === 'pell' ? 1 : 2);
        break;
      case 'DEC_JUMP':
        payload.push(instr.reg === 'phi' ? 3 : instr.reg === 'pell' ? 4 : 5);
        payload.push(instr.target);
        break;
      case 'RACE':
        payload.push(6);
        payload.push(instr.k);
        break;
      case 'HALT':
        payload.push(7);
        break;
    }
  }
  return encodeZeckendorfArray(payload);
}

/**
 * Handles the gnosis decode UPMProgram workflow.
 */
export function decodeUPMProgram(bytes: Uint8Array): UPMInstruction[] {
  const payload = decodeZeckendorfArray(bytes);
  const program: UPMInstruction[] = [];
  for (let i = 0; i < payload.length; i++) {
    const opcode = payload[i]!;
    if (opcode === 0) program.push({ op: 'INC', reg: 'phi' });
    else if (opcode === 1) program.push({ op: 'INC', reg: 'pell' });
    else if (opcode === 2) program.push({ op: 'INC', reg: 'trib' });
    else if (opcode >= 3 && opcode <= 5) {
      const target = payload[++i]!;
      const reg = opcode === 3 ? 'phi' : opcode === 4 ? 'pell' : 'trib';
      program.push({ op: 'DEC_JUMP', reg, target });
    }
    else if (opcode === 6) {
      program.push({ op: 'RACE', k: payload[++i]! });
    }
    else if (opcode === 7) {
      program.push({ op: 'HALT' });
    }
  }
  return program;
}
