// @a0n/.ignored_gnosis/src/verify-bundle.ts
/**
 * Deterministic verification bundle checker (Arena 5).
 * No network: pure JSON structural validation + policy predicates.
 */

export interface VerificationGate {
  readonly name: string;
  /** Must be 0 for green */
  readonly exit: number;
  readonly log_hash: string;
}

export interface VerificationFormal {
  readonly sorry_count_delta?: number;
  readonly touched_modules?: readonly string[];
}

export interface VerificationBundleV1 {
  readonly schemaVersion: '1.0.0';
  readonly patch_content_id: string;
  readonly toolchain_id: string;
  readonly ci_config_id: string;
  readonly gates: readonly VerificationGate[];
  readonly formal?: VerificationFormal;
  readonly signatures?: readonly { format: 'sigstore' | 'none'; payload?: string }[];
}

export type VerifyResult =
  | { ok: true }
  | { ok: false; reason: string };

/**
 * `verify(bundle) ∈ {⊤,⊥}` — deterministic, no I/O.
 */
export function verifyBundle(bundle: unknown): VerifyResult {
  if (bundle === null || typeof bundle !== 'object') {
    return { ok: false, reason: 'bundle_not_object' };
  }
  const b = bundle as Record<string, unknown>;
  if (b.schemaVersion !== '1.0.0') {
    return { ok: false, reason: 'schema_version' };
  }
  for (const k of ['patch_content_id', 'toolchain_id', 'ci_config_id'] as const) {
    const v = b[k];
    if (typeof v !== 'string' || v.length < 8) {
      return { ok: false, reason: `field_${k}` };
    }
  }
  const gates = b.gates;
  if (!Array.isArray(gates) || gates.length < 1) {
    return { ok: false, reason: 'gates' };
  }
  for (let i = 0; i < gates.length; i++) {
    const g = gates[i] as Record<string, unknown>;
    if (typeof g.name !== 'string' || g.name.length < 1) {
      return { ok: false, reason: `gate_${i}_name` };
    }
    if (g.exit !== 0) {
      return { ok: false, reason: `gate_${i}_exit` };
    }
    if (typeof g.log_hash !== 'string' || g.log_hash.length < 8) {
      return { ok: false, reason: `gate_${i}_log_hash` };
    }
  }
  const formal = b.formal as Record<string, unknown> | undefined;
  if (formal !== undefined) {
    const d = formal.sorry_count_delta;
    if (d !== undefined && typeof d === 'number' && d > 0) {
      return { ok: false, reason: 'sorry_count_delta' };
    }
  }
  return { ok: true };
}
