// @a0n/.ignored_gnosis/src/void-agent.ts
/**
 * void-agent.ts -- METACOG: the agent primitive
 *
 * An agent maps to a void walk. Not "has" a personality -- formalizes the personality.
 * Not "runs" an attention mechanism -- formalizes the attention.
 *
 * The VoidAgent composes the primitives that gnosis has been building:
 *   - BoundaryStack (personality): 7-layer timescale hierarchy
 *   - VoidAttentionHead[] (perception): multi-head attention over action space
 *   - VoidCrossAttentionHead (social): attention to other agents' boundaries
 *   - Walker (metacognition): c0-c3 loop formalizes the agent loop
 *
 * The agent loop maps exactly to the metacognitive cycle:
 *   perceive  = attend (complement distribution over stimuli)
 *   decide    = c0Choose (sample from personality-constrained complement)
 *   act       = emit the chosen action
 *   observe   = c0Update (rejection accumulates as void)
 *   reflect   = c1Measure (entropy, kurtosis, gini, inverseBule)
 *   adapt     = c2c3 (gait determines depth of future deliberation)
 *
 * Personality constrains everything. Temperament (lifetime boundary) shapes
 * baseline eta. Attachment (lifetime) biases social attention. Traits (years)
 * are the long-term complement shape. Behaviors (months) are gait history.
 * Mental health (weeks) is the kurtosis. History is the boundary itself.
 *
 * The agent doesn't "have" a personality. The agent formalizes the walk.
 */

import {
//   type _VoidBoundary,
//   type _Walker,
  type Gait,
  type Measurement,
  type BoundaryStack,
//   type TimescaleBoundary,
  type StackWalker,
  type TimescaleBoundary,
  type Resonance,
//   createVoidBoundary,
//   createWalker,
  createTimescaleBoundary,
  createBoundaryStack,
  createStackWalker,
  createResonance,
  complementDistribution,
  sampleComplement,
  updateVoidBoundary,
//   measure,
//   measureStack,
  flattenStack,
  stepStackWalker,
  measureStackWalker,
  upwardConstraint,
//   c0Choose,
//   c0Update,
  c1Measure,
//   c2c3Adapt,
//   projectBoundary,
  GAIT_DEPTH,
} from './void';

import {
  type VoidAttentionHead,
  type VoidCrossAttentionHead,
  createAttentionHead,
  createCrossAttentionHead,
  attend,
  reject as rejectAttention,
  layerNorm as attentionLayerNorm,
  feedForward as attentionFeedForward,
  crossAttend,
  crossReject,
  crossLayerNorm,
  crossFeedForward,
} from './void-attention';

// ============================================================================
// Agent Configuration
// ============================================================================

export interface VoidAgentConfig {
  /** Human-readable name */
  name: string;
  /** Dimensionality of the action space */
  actionDimensions: number;
  /** Number of attention heads for perception */
  numHeads?: number;
  /** Initial temperature */
  eta?: number;
  /** Neighborhood radius for rejection spread */
  neighborhoodRadius?: number;
  /** Decay rate per tick (layer norm) */
  decayRate?: number;
  /** Optional personality stack (if not provided, a flat boundary is used) */
  personality?: BoundaryStack;
  /** Optional pre-initialized personality layers */
  personalityLayers?: PersonalityLayerConfig[];
}

/**
 * Configuration for a personality layer.
 * Maps to the 7-layer model from aeon-affectively:
 *   1. Temperament (lifetime)     -- birth configuration
 *   2. Attachment  (lifetime)     -- singularity patterns
 *   3. Traits      (years)        -- complement shape
 *   4. Behaviors   (months)       -- gait selection
 *   5. Mental Health (weeks)      -- kurtosis
 *   6. History     (years)        -- the boundary itself
 *   7. Culture     (generational) -- inherited void
 */
export interface PersonalityLayerConfig {
  name: string;
  timescale:
    | 'instant'
    | 'seconds'
    | 'minutes'
    | 'hours'
    | 'days'
    | 'weeks'
    | 'months'
    | 'years'
    | 'lifetime'
    | 'generational';
  dimensions: number;
  /** Optional dimension labels for introspection */
  labels?: readonly string[];
  /** Optional initial void counts (pre-accumulated experience) */
  initialCounts?: number[];
}

// ============================================================================
// VoidAgent -- the primitive
// ============================================================================

export interface VoidAgent {
  /** Human-readable name */
  name: string;

  /** Personality: multi-timescale boundary stack */
  personality: StackWalker;

  /** Perception: multi-head attention over action space */
  perception: VoidAttentionHead[];

  /** Social: cross-attention to other agents */
  social: VoidCrossAttentionHead | null;

  /** Action space dimensionality */
  actionDimensions: number;

  /** RNG source */
  rng: () => number;

  /** Total ticks executed */
  ticks: number;

  /** Action history */
  history: AgentTick[];
}

export interface AgentTick {
  tick: number;
  /** What the agent perceived (attention weights) */
  perception: number[];
  /** What the agent chose */
  action: number;
  /** The gait at decision time */
  gait: Gait;
  /** Whether the action was rejected by the environment */
  rejected: boolean;
  /** Rejection magnitude */
  rejectionMagnitude: number;
  /** Reward received */
  reward: number;
  /** Self-measurement after the tick */
  measurement: Measurement;
  /** Personality-level measurement */
  personalityMeasurement: {
    layers: Measurement[];
    global: Measurement;
    gaits: Gait[];
  };
}

function requirePrimaryPerceptionHead(
  agent: VoidAgent,
  context: string
): VoidAttentionHead {
  const head = agent.perception[0];
  if (head === undefined) {
    throw new Error(`${context}: missing primary perception head`);
  }
  return head;
}

function requirePrimaryPersonalityLayer(
  agent: VoidAgent,
  context: string
): TimescaleBoundary {
  const layer = agent.personality.stack.layers[0];
  if (layer === undefined) {
    throw new Error(`${context}: missing primary personality layer`);
  }
  return layer;
}

// ============================================================================
// Creation
// ============================================================================

/**
 * Creates the Void Agent.
 */
export function createVoidAgent(
  config: VoidAgentConfig,
  rng: () => number = Math.random
): VoidAgent {
  const eta = config.eta ?? 2.0;
  const numHeads = config.numHeads ?? 1;
  const neighborhoodRadius = config.neighborhoodRadius ?? 1;
  const decayRate = config.decayRate ?? 0.01;

  // Build personality stack
  let personality: StackWalker;
  if (config.personality) {
    personality = createStackWalker(config.personality);
  } else if (config.personalityLayers && config.personalityLayers.length > 0) {
    const layers = config.personalityLayers.map((lc) => {
      const tb = createTimescaleBoundary(
        lc.name,
        lc.timescale,
        lc.dimensions,
        lc.labels
      );
      if (lc.initialCounts) {
        for (let i = 0; i < lc.initialCounts.length && i < lc.dimensions; i++) {
          const initialCount = lc.initialCounts[i];
          if (initialCount !== undefined && initialCount > 0) {
            updateVoidBoundary(tb.boundary, i, initialCount);
          }
        }
      }
      return tb;
    });
    // Default resonance: temperament <-> mental health, culture <-> attachment
    const resonances: Resonance[] = [];
    if (layers.length >= 5) {
      resonances.push(
        createResonance(0, 4, 0.05, 'temperament <-> mental health')
      );
    }
    if (layers.length >= 7) {
      resonances.push(createResonance(6, 1, 0.05, 'culture <-> attachment'));
    }
    const stack = createBoundaryStack(
      `${config.name}:personality`,
      layers,
      resonances
    );
    personality = createStackWalker(stack);
  } else {
    // Minimal: single-layer personality at the action space dimensionality
    const layer = createTimescaleBoundary(
      'flat',
      'days',
      config.actionDimensions
    );
    const stack = createBoundaryStack(`${config.name}:flat`, [layer]);
    personality = createStackWalker(stack);
  }

  // Build perception heads
  const perception = Array.from({ length: numHeads }, () =>
    createAttentionHead(
      config.actionDimensions,
      eta,
      neighborhoodRadius,
      decayRate
    )
  );

  return {
    name: config.name,
    personality,
    perception,
    social: null,
    actionDimensions: config.actionDimensions,
    rng,
    ticks: 0,
    history: [],
  };
}

// ============================================================================
// Social bonding -- cross-attention to another agent
// ============================================================================

/**
 * Bond two agents: create cross-attention heads so each can attend
 * to the other's action boundary. This is how agents perceive each other.
 */
export function bond(
  agentA: VoidAgent,
  agentB: VoidAgent,
  eta?: number,
  neighborhoodRadius?: number
): void {
  const agentAPrimary = requirePrimaryPerceptionHead(agentA, 'bond');
  const agentBPrimary = requirePrimaryPerceptionHead(agentB, 'bond');
  agentA.social = createCrossAttentionHead(
    agentA.actionDimensions,
    agentB.actionDimensions,
    eta ?? agentAPrimary.walker.eta,
    neighborhoodRadius ?? agentAPrimary.neighborhoodRadius,
    agentAPrimary.decayRate
  );
  agentB.social = createCrossAttentionHead(
    agentB.actionDimensions,
    agentA.actionDimensions,
    eta ?? agentBPrimary.walker.eta,
    neighborhoodRadius ?? agentBPrimary.neighborhoodRadius,
    agentBPrimary.decayRate
  );
}

// ============================================================================
// The agent loop: perceive → decide → act → observe → reflect → adapt
// ============================================================================

/**
 * Perceive: compute attention weights over the action space.
 * Personality constrains perception -- deeper layers bias the
 * complement distribution through upward constraint.
 */
export function perceive(agent: VoidAgent): {
  weights: number[];
  perHead: { weights: number[]; selected: number; measurement: Measurement }[];
} {
  // Apply personality constraint to perception boundaries
  applyPersonalityConstraint(agent);

  // Multi-head attention
  const perHead = agent.perception.map((h) => attend(h, agent.rng));

  // Merge heads: average weights
  const merged = new Array(agent.actionDimensions).fill(0);
  for (const head of perHead) {
    for (let i = 0; i < head.weights.length; i++) {
      merged[i] += head.weights[i];
    }
  }
  const numHeads = perHead.length;
  for (let i = 0; i < merged.length; i++) merged[i] /= numHeads;

  return { weights: merged, perHead };
}

/**
 * Perceive another agent through social cross-attention.
 * Returns null if no social bond exists.
 */
export function perceiveOther(
  agent: VoidAgent,
  other: VoidAgent
): {
  weights: number[];
  selected: [number, number];
  measurement: Measurement;
} | null {
  if (!agent.social) return null;
  const agentPrimary = requirePrimaryPerceptionHead(agent, 'perceiveOther');
  const otherPrimary = requirePrimaryPerceptionHead(other, 'perceiveOther');
  return crossAttend(
    agent.social,
    agentPrimary.walker.boundary,
    otherPrimary.walker.boundary,
    agentPrimary.walker.eta,
    otherPrimary.walker.eta,
    agent.rng
  );
}

/**
 * Decide: choose an action from the personality-constrained complement.
 *
 * The decision is a two-stage process:
 * 1. Perception provides attention weights over actions
 * 2. Personality biases those weights through upward constraint
 *
 * If social attention is active, the other agent's boundary also
 * influences the decision (gated by cross-attention).
 */
export function decide(
  agent: VoidAgent,
  perceptionWeights: number[],
  socialInfluence?: { selected: [number, number] } | null
): number {
  // Primary: sample from perception-weighted complement
  const ownChoice = sampleComplement(perceptionWeights, agent.rng);

  // Social: if cross-attention suggests an action, accept if its
  // perception weight exceeds our own choice's weight
  if (socialInfluence) {
    const proposed = socialInfluence.selected[0]; // our-dimension of the proposal
    const proposedWeight = proposed === undefined ? undefined : perceptionWeights[proposed];
    const ownWeight = perceptionWeights[ownChoice];
    if (
      proposed !== undefined &&
      proposed < perceptionWeights.length &&
      proposedWeight !== undefined &&
      ownWeight !== undefined &&
      proposedWeight >= ownWeight
    ) {
      return proposed;
    }
  }

  return ownChoice;
}

/**
 * Observe: the environment returns an outcome. Rejection accumulates void.
 *
 * The rejection predicate is domain-specific -- the caller decides
 * what constitutes rejection. The agent just accumulates.
 */
export function observe(
  agent: VoidAgent,
  action: number,
  rejected: boolean,
  rejectionMagnitude = 1,
  reward = 0
): void {
  // Update perception heads
  if (rejected) {
    for (const head of agent.perception) {
      rejectAttention(head, action, rejectionMagnitude);
    }
  } else {
    // Even success accumulates a small amount of void (mild habituation)
    for (const head of agent.perception) {
      rejectAttention(head, action, 0.1);
    }
  }

  // Update social cross-attention if bonded
  if (agent.social && rejected) {
    // We don't know the other agent's action here, so we reject
    // at our own action dimension across all of theirs
    for (let j = 0; j < agent.social.dimB; j++) {
      crossReject(
        agent.social,
        action,
        j,
        Math.max(1, Math.round(rejectionMagnitude / (j + 1)))
      );
    }
  }

  // Propagate to personality stack
  propagateToPersonality(agent, action, rejected, rejectionMagnitude, reward);
}

/**
 * Reflect: the agent measures its own state.
 * This is c1Measure across all systems.
 */
export function reflect(agent: VoidAgent): {
  perception: Measurement;
  personality: { layers: Measurement[]; global: Measurement; gaits: Gait[] };
  social: Measurement | null;
  gait: Gait;
} {
  const primary = requirePrimaryPerceptionHead(agent, 'reflect');
  const perceptionMeasurement = c1Measure(primary.walker);
  const personalityMeasurement = measureStackWalker(agent.personality);
  const socialMeasurement = agent.social
    ? c1Measure(agent.social.walker)
    : null;

  return {
    perception: perceptionMeasurement,
    personality: personalityMeasurement,
    social: socialMeasurement,
    gait: primary.walker.gait,
  };
}

/**
 * Adapt: the agent adjusts its parameters.
 * This is c2c3Adapt + layer norm + personality tick.
 */
export function adapt(agent: VoidAgent): void {
  // Perception: layer norm + feed-forward
  for (const head of agent.perception) {
    attentionLayerNorm(head);
    attentionFeedForward(head);
  }

  // Social: layer norm + feed-forward
  if (agent.social) {
    crossLayerNorm(agent.social);
    crossFeedForward(agent.social);
  }

  // Personality stack: tick (natural decay + inter-layer propagation)
  stepStackWalker(agent.personality, []);
}

// ============================================================================
// Full tick: the complete agent loop
// ============================================================================

/**
 * Execute one full metacognitive tick:
 *   perceive → decide → (act is external) → observe → reflect → adapt
 *
 * Returns the chosen action. The caller must execute the action in
 * the environment and call observe() with the outcome.
 */
export function tick(
  agent: VoidAgent,
  other?: VoidAgent | null
): { action: number; perception: number[] } {
  // Perceive
  const p = perceive(agent);

  // Social perception
  const socialInfluence = other ? perceiveOther(agent, other) : null;

  // Decide
  const action = decide(agent, p.weights, socialInfluence);

  agent.ticks++;

  return { action, perception: p.weights };
}

/**
 * Complete a tick after the environment responds.
 * Call this after tick() with the environment's feedback.
 */
export function completeTick(
  agent: VoidAgent,
  action: number,
  perception: number[],
  rejected: boolean,
  rejectionMagnitude = 1,
  reward = 0
): AgentTick {
  // Observe
  observe(agent, action, rejected, rejectionMagnitude, reward);

  // Reflect
  const r = reflect(agent);

  // Adapt
  adapt(agent);

  const tickResult: AgentTick = {
    tick: agent.ticks,
    perception,
    action,
    gait: r.gait,
    rejected,
    rejectionMagnitude,
    reward,
    measurement: r.perception,
    personalityMeasurement: r.personality,
  };

  agent.history.push(tickResult);
  return tickResult;
}

/**
 * Run a full perceive-decide-observe-reflect-adapt cycle in one call.
 * For when the environment response is immediate (synchronous games, etc).
 */
export function step(
  agent: VoidAgent,
  /** Environment: given agent's action, returns [rejected, magnitude, reward] */
  environment: (action: number) => {
    rejected: boolean;
    magnitude?: number;
    reward?: number;
  },
  other?: VoidAgent | null
): AgentTick {
  const { action, perception } = tick(agent, other);
  const outcome = environment(action);
  return completeTick(
    agent,
    action,
    perception,
    outcome.rejected,
    outcome.magnitude ?? 1,
    outcome.reward ?? 0
  );
}

// ============================================================================
// Personality ↔ Perception coupling
// ============================================================================

/**
 * Apply personality constraint to perception.
 * The deepest personality layer's complement distribution biases
 * the perception heads' boundaries -- where personality says "don't go",
 * perception accumulates void before attending.
 */
function applyPersonalityConstraint(agent: VoidAgent): void {
  const layers = agent.personality.stack.layers;
  if (layers.length === 0) return;

  // Use the deepest layer (temperament/inherited) as the primary constraint
  const deepest = requirePrimaryPersonalityLayer(agent, 'applyPersonalityConstraint');
  const primary = requirePrimaryPerceptionHead(agent, 'applyPersonalityConstraint');
  const constraint = upwardConstraint(
    deepest,
    {
      boundary: primary.walker.boundary,
      timescale: 'instant',
      name: 'perception',
    },
    0.05
  );

  // Apply constraint to all perception heads
  for (const head of agent.perception) {
    const dims = Math.min(
      constraint.length,
      head.walker.boundary.counts.length
    );
    for (let i = 0; i < dims; i++) {
      const current = head.walker.boundary.counts[i] ?? 0;
      const delta = constraint[i] ?? 0;
      head.walker.boundary.counts[i] = current + delta;
    }
    head.walker.boundary.totalEntries = head.walker.boundary.counts.reduce(
      (a: number, b: number) => a + b,
      0
    );
  }
}

/**
 * Propagate action outcomes to the personality stack.
 * Different personality layers absorb the signal at different magnitudes:
 *   - Behaviors (layer 4, months): full signal
 *   - Mental health (layer 5, weeks): amplified for rejection
 *   - Traits (layer 3, years): attenuated signal
 *   - Deeper layers: minimal (they change slowly)
 */
function propagateToPersonality(
  agent: VoidAgent,
  action: number,
  rejected: boolean,
  magnitude: number,
  reward: number
): void {
  const layers = agent.personality.stack.layers;
  const rejections: {
    layerIdx: number;
    dimensionIdx: number;
    magnitude: number;
    reward?: number;
  }[] = [];

  for (let i = 0; i < layers.length; i++) {
    const layer = layers[i];
    if (layer === undefined) {
      continue;
    }
    const dims = layer.boundary.counts.length;

    // Map action to this layer's dimensionality
    const dimIdx = action % dims;

    // Magnitude attenuation by timescale depth
    let layerMag: number;
    switch (layer.timescale) {
      case 'instant':
      case 'seconds':
      case 'minutes':
        layerMag = magnitude; // Full signal for fast layers
        break;
      case 'hours':
      case 'days':
        layerMag = magnitude * 0.5; // Moderate for medium layers
        break;
      case 'weeks':
        // Mental health amplifies rejection
        layerMag = rejected ? magnitude * 1.5 : magnitude * 0.3;
        break;
      case 'months':
        // Behaviors absorb the full signal
        layerMag = magnitude * 0.8;
        break;
      case 'years':
        layerMag = magnitude * 0.1; // Traits change slowly
        break;
      case 'lifetime':
      case 'generational':
        layerMag = magnitude * 0.01; // Nearly immutable
        break;
      default:
        layerMag = magnitude * 0.1;
    }

    if (rejected && layerMag > 0) {
      rejections.push({
        layerIdx: i,
        dimensionIdx: dimIdx,
        magnitude: layerMag,
        reward,
      });
    } else if (!rejected) {
      // Success still leaves a trace (mild habituation) at fast layers
      if (
        layer.timescale === 'instant' ||
        layer.timescale === 'seconds' ||
        layer.timescale === 'minutes'
      ) {
        rejections.push({
          layerIdx: i,
          dimensionIdx: dimIdx,
          magnitude: 0.05,
          reward,
        });
      }
    }
  }

  stepStackWalker(agent.personality, rejections);
}

// ============================================================================
// Introspection -- the agent reads itself
// ============================================================================

/**
 * The agent's current personality profile as a single complement distribution.
 * This is the flattened stack -- the full state vector projected through void.
 */
export function personalityVector(agent: VoidAgent, eta?: number): number[] {
  const flat = flattenStack(agent.personality.stack);
  const primary = requirePrimaryPerceptionHead(agent, 'personalityVector');
  return complementDistribution(flat, eta ?? primary.walker.eta);
}

/**
 * Which actions is the agent most likely to take right now?
 * Returns actions ranked by complement weight (freshest first).
 */
export function actionPreferences(
  agent: VoidAgent
): { action: number; weight: number }[] {
  const primary = requirePrimaryPerceptionHead(agent, 'actionPreferences');
  const dist = complementDistribution(primary.walker.boundary, primary.walker.eta);
  return dist
    .map((weight: number, action: number) => ({ action, weight }))
    .sort(
      (
        a: { action: number; weight: number },
        b: { action: number; weight: number }
      ) => b.weight - a.weight
    );
}

/**
 * What has the agent rejected most?
 * Returns actions ranked by void accumulation (most rejected first).
 */
export function rejectionProfile(
  agent: VoidAgent
): { action: number; voidCount: number }[] {
  const primary = requirePrimaryPerceptionHead(agent, 'rejectionProfile');
  const counts = primary.walker.boundary.counts;
  return counts
    .map((voidCount: number, action: number) => ({ action, voidCount }))
    .sort(
      (
        a: { action: number; voidCount: number },
        b: { action: number; voidCount: number }
      ) => b.voidCount - a.voidCount
    );
}

/**
 * The agent's metacognitive state: am I exploring or exploiting?
 * Am I rigid or flexible? Am I converging or diverging?
 */
export function metacogState(agent: VoidAgent): {
  gait: Gait;
  depth: number;
  eta: number;
  exploration: number;
  entropy: number;
  kurtosis: number;
  inverseBule: number;
  steps: number;
  adaptations: number;
} {
  const primary = requirePrimaryPerceptionHead(agent, 'metacogState');
  const walker = primary.walker;
  const m = c1Measure(walker);
  return {
    gait: walker.gait,
    depth: GAIT_DEPTH[walker.gait as Gait],
    eta: walker.eta,
    exploration: walker.exploration,
    entropy: m.entropy,
    kurtosis: m.kurtosis,
    inverseBule: m.inverseBule,
    steps: walker.steps,
    adaptations: walker.adaptations,
  };
}
