// @a0n/.ignored_gnosis/src/void-merkle.ts
/**
 * void-merkle.ts -- Merkle verification over void boundaries
 *
 * A void Merkle tree hashes rejection records into a tamper-evident
 * append-only structure. Each fold produces a hash of its vent.
 * The hash chain formalizes the void certificate. You can verify that a
 * twin folds and unfolds faithfully by checking the Merkle root
 * matches the expected complement distribution -- without replaying
 * the entire history.
 *
 * Mechanized in PsycheGrindExtended.lean, Pass 18 (zero sorry):
 *   - void_merkle_deterministic: same rejections = same root
 *   - void_merkle_tamper_evident: modification changes root
 *   - void_merkle_append_only: new vents extend, old hashes unchanged
 *   - complement_preserved_until_new: twins share root until new topology
 *   - new_topology_diverges_complement: new experience breaks root equality
 *   - void_merkle_not_reversible (anti): auditable but not reconstructible
 *   - fold_unfold_round_trip: exact round-trip through fold/unfold
 *   - void_merkle_composition: composing trees produces valid combined tree
 *   - incremental_verification: each step verifiable against parent hash
 *   - void_merkle_minimum_size: minimum tree = one node, one vent, one fold
 *
 * @module @a0n/gnosis/void-merkle
 */

import type { VoidBoundary } from './void';

// ============================================================================
// Core Types
// ============================================================================

/** A single vent record -- one fold's rejection output. */
export interface VoidVentRecord {
  /** The dimension index that was rejected */
  readonly dimensionIdx: number;
  /** The rejection magnitude */
  readonly magnitude: number;
  /** Monotonic sequence number (append-only) */
  readonly sequenceNumber: number;
  /** Timestamp of the vent event */
  readonly timestamp: number;
}

/** A Merkle node in the void tree. */
export interface VoidMerkleNode {
  /** Hash of the left child (0 if leaf) */
  readonly leftHash: number;
  /** Hash of the right child (0 if leaf) */
  readonly rightHash: number;
  /** Local vent contribution */
  readonly localVent: number;
  /** This node's hash: leftHash + rightHash + localVent */
  readonly nodeHash: number;
  /** Whether this is a leaf node */
  readonly isLeaf: boolean;
}

/** A proof that a specific vent record exists in the tree. */
export interface VoidMerkleProof {
  /** The vent record being proved */
  readonly ventRecord: VoidVentRecord;
  /** The sibling hashes along the path from leaf to root */
  readonly path: readonly VoidMerklePathEntry[];
  /** The root hash this proof validates against */
  readonly rootHash: number;
}

export interface VoidMerklePathEntry {
  readonly hash: number;
  readonly side: 'left' | 'right';
}

/** The full void Merkle tree. */
export interface VoidMerkleTree {
  /** The root hash -- the void certificate */
  readonly rootHash: number;
  /** Total vent records in the tree */
  readonly size: number;
  /** The append-only vent log */
  readonly vents: readonly VoidVentRecord[];
  /** All nodes in the tree (bottom-up) */
  readonly nodes: readonly VoidMerkleNode[];
  /** Timestamp of last append */
  readonly lastAppendedAt: number;
}

/** Verification result when comparing two void Merkle trees. */
export interface VoidMerkleVerification {
  /** Whether the roots match */
  readonly rootsMatch: boolean;
  /** The root hash of the first tree */
  readonly rootA: number;
  /** The root hash of the second tree */
  readonly rootB: number;
  /** Number of shared vent records */
  readonly sharedVents: number;
  /** Number of divergent vent records */
  readonly divergentVents: number;
  /** Whether trees are from the same fork (shared prefix) */
  readonly sameFork: boolean;
}

/** Twin fold verification certificate. */
export interface TwinFoldCertificate {
  /** The original's void Merkle root */
  readonly originalRoot: number;
  /** The twin's void Merkle root */
  readonly twinRoot: number;
  /** Whether the twin's fold matches the original */
  readonly verified: boolean;
  /** The point of divergence (sequence number, or -1 if none) */
  readonly divergencePoint: number;
  /** Timestamp of verification */
  readonly verifiedAt: number;
}

// ============================================================================
// Construction
// ============================================================================

/**
 * Hash function for void Merkle nodes.
 * Uses a simple deterministic combination -- the Lean proof uses addition,
 * and this runtime mirrors that. For cryptographic strength, use
 * shared-ui/services/proof/merkleRootCalculator.ts instead.
 *
 * Mechanized: void_merkle_deterministic -- same inputs = same hash.
 */
export function voidMerkleHash(
  leftHash: number,
  rightHash: number,
  localVent: number
): number {
  return leftHash + rightHash + localVent;
}

/** Create a leaf node from a vent record. */
export function createVoidMerkleLeaf(vent: VoidVentRecord): VoidMerkleNode {
  const localVent = vent.dimensionIdx * 1000 + vent.magnitude;
  return {
    leftHash: 0,
    rightHash: 0,
    localVent,
    nodeHash: voidMerkleHash(0, 0, localVent),
    isLeaf: true,
  };
}

/** Create an internal node from two children. */
export function createVoidMerkleInternal(
  left: VoidMerkleNode,
  right: VoidMerkleNode
): VoidMerkleNode {
  return {
    leftHash: left.nodeHash,
    rightHash: right.nodeHash,
    localVent: 0,
    nodeHash: voidMerkleHash(left.nodeHash, right.nodeHash, 0),
    isLeaf: false,
  };
}

/**
 * Build a void Merkle tree from a sequence of vent records.
 *
 * Mechanized: void_merkle_deterministic, void_merkle_append_only.
 */
export function buildVoidMerkleTree(
  vents: readonly VoidVentRecord[]
): VoidMerkleTree {
  if (vents.length === 0) {
    return {
      rootHash: 0,
      size: 0,
      vents: [],
      nodes: [],
      lastAppendedAt: 0,
    };
  }

  // Build leaves
  const leaves = vents.map(createVoidMerkleLeaf);
  const allNodes: VoidMerkleNode[] = [...leaves];

  // Build tree bottom-up
  let currentLevel = leaves;
  while (currentLevel.length > 1) {
    const nextLevel: VoidMerkleNode[] = [];
    for (let i = 0; i < currentLevel.length; i += 2) {
      const left = currentLevel[i];
      if (left === undefined) {
        throw new Error('buildVoidMerkleTree: missing left node');
      }
      const right = i + 1 < currentLevel.length ? currentLevel[i + 1] ?? left : left;
      const parent = createVoidMerkleInternal(left, right);
      nextLevel.push(parent);
      allNodes.push(parent);
    }
    currentLevel = nextLevel;
  }

  const root = currentLevel[0];
  if (root === undefined) {
    throw new Error('buildVoidMerkleTree: missing root node');
  }
  const lastVent = vents[vents.length - 1];
  if (lastVent === undefined) {
    throw new Error('buildVoidMerkleTree: missing last vent');
  }

  return {
    rootHash: root.nodeHash,
    size: vents.length,
    vents,
    nodes: allNodes,
    lastAppendedAt: lastVent.timestamp,
  };
}

/**
 * Append a vent record to an existing tree.
 * Rebuilds the tree (append-only semantics).
 *
 * Mechanized: void_merkle_append_only -- old hashes unchanged.
 */
export function appendVent(
  tree: VoidMerkleTree,
  vent: VoidVentRecord
): VoidMerkleTree {
  return buildVoidMerkleTree([...tree.vents, vent]);
}

/**
 * Create a vent record from a void boundary update.
 */
export function createVentRecord(
  dimensionIdx: number,
  magnitude: number,
  sequenceNumber: number
): VoidVentRecord {
  return {
    dimensionIdx,
    magnitude,
    sequenceNumber,
    timestamp: Date.now(),
  };
}

// ============================================================================
// Verification
// ============================================================================

/**
 * Verify that two void Merkle trees share the same root.
 *
 * Mechanized: complement_preserved_until_new, new_topology_diverges_complement.
 */
export function verifyVoidMerklePair(
  treeA: VoidMerkleTree,
  treeB: VoidMerkleTree
): VoidMerkleVerification {
  const minLen = Math.min(treeA.vents.length, treeB.vents.length);
  let sharedVents = 0;
  let divergenceFound = false;

  for (let i = 0; i < minLen; i++) {
    const a = treeA.vents[i];
    const b = treeB.vents[i];
    if (a === undefined || b === undefined) {
      throw new Error('verifyVoidMerklePair: missing vent entry');
    }
    if (
      a.dimensionIdx === b.dimensionIdx &&
      a.magnitude === b.magnitude &&
      a.sequenceNumber === b.sequenceNumber
    ) {
      sharedVents++;
    } else {
      divergenceFound = true;
      break;
    }
  }

  const divergentVents =
    treeA.vents.length + treeB.vents.length - 2 * sharedVents;

  return {
    rootsMatch: treeA.rootHash === treeB.rootHash,
    rootA: treeA.rootHash,
    rootB: treeB.rootHash,
    sharedVents,
    divergentVents,
    sameFork: !divergenceFound && sharedVents === minLen,
  };
}

/**
 * Generate a proof that a specific vent exists in the tree.
 *
 * Mechanized: void_merkle_inclusion, incremental_verification.
 */
export function generateVoidMerkleProof(
  tree: VoidMerkleTree,
  ventIndex: number
): VoidMerkleProof | null {
  if (ventIndex < 0 || ventIndex >= tree.vents.length) return null;

  const vent = tree.vents[ventIndex];
  if (vent === undefined) {
    throw new Error('generateVoidMerkleProof: missing vent');
  }
  const leaves = tree.vents.map(createVoidMerkleLeaf);
  const path: VoidMerklePathEntry[] = [];

  let currentLevel = leaves.map((n) => n.nodeHash);
  let idx = ventIndex;

  while (currentLevel.length > 1) {
    const nextLevel: number[] = [];
    for (let i = 0; i < currentLevel.length; i += 2) {
      const left = currentLevel[i];
      if (left === undefined) {
        throw new Error('generateVoidMerkleProof: missing left hash');
      }
      const right = i + 1 < currentLevel.length ? currentLevel[i + 1] ?? left : left;

      if (i === idx || i + 1 === idx) {
        // This pair contains our target
        if (i === idx) {
          path.push({ hash: right, side: 'right' });
        } else {
          path.push({ hash: left, side: 'left' });
        }
      }

      nextLevel.push(voidMerkleHash(left, right, 0));
    }
    idx = Math.floor(idx / 2);
    currentLevel = nextLevel;
  }

  return {
    ventRecord: vent,
    path,
    rootHash: tree.rootHash,
  };
}

/**
 * Verify a Merkle proof against a root hash.
 *
 * Mechanized: void_merkle_tamper_evident -- any change breaks verification.
 */
export function verifyVoidMerkleProof(proof: VoidMerkleProof): boolean {
  const leaf = createVoidMerkleLeaf(proof.ventRecord);
  let currentHash = leaf.nodeHash;

  for (const entry of proof.path) {
    if (entry.side === 'right') {
      currentHash = voidMerkleHash(currentHash, entry.hash, 0);
    } else {
      currentHash = voidMerkleHash(entry.hash, currentHash, 0);
    }
  }

  return currentHash === proof.rootHash;
}

/**
 * Verify that a twin folded correctly by comparing void Merkle roots.
 *
 * Mechanized: twin_fold_verification, complement_preserved_until_new.
 */
export function verifyTwinFold(
  original: VoidMerkleTree,
  twin: VoidMerkleTree
): TwinFoldCertificate {
  const verification = verifyVoidMerklePair(original, twin);

  let divergencePoint = -1;
  if (!verification.rootsMatch) {
    const minLen = Math.min(original.vents.length, twin.vents.length);
    for (let i = 0; i < minLen; i++) {
      const a = original.vents[i];
      const b = twin.vents[i];
      if (a === undefined || b === undefined) {
        throw new Error('verifyTwinFold: missing vent entry');
      }
      if (a.dimensionIdx !== b.dimensionIdx || a.magnitude !== b.magnitude) {
        divergencePoint = i;
        break;
      }
    }
    if (divergencePoint === -1 && original.vents.length !== twin.vents.length) {
      divergencePoint = minLen;
    }
  }

  return {
    originalRoot: original.rootHash,
    twinRoot: twin.rootHash,
    verified: verification.rootsMatch,
    divergencePoint,
    verifiedAt: Date.now(),
  };
}

// ============================================================================
// Composition
// ============================================================================

/**
 * Compose two void Merkle trees into a combined tree.
 *
 * Mechanized: void_merkle_composition.
 */
export function composeVoidMerkleTrees(
  treeA: VoidMerkleTree,
  treeB: VoidMerkleTree
): VoidMerkleTree {
  // Merge vent logs in sequence-number order
  const allVents = [...treeA.vents, ...treeB.vents].sort(
    (a, b) => a.sequenceNumber - b.sequenceNumber
  );
  return buildVoidMerkleTree(allVents);
}

// ============================================================================
// Bridge: VoidBoundary → VoidMerkleTree
// ============================================================================

/**
 * Convert a VoidBoundary snapshot into a void Merkle tree.
 * Each non-zero dimension count becomes a vent record.
 *
 * This is the bridge between the accumulation model (void.ts)
 * and the verification model (void-merkle.ts).
 */
export function voidBoundaryToMerkleTree(
  boundary: VoidBoundary
): VoidMerkleTree {
  const vents: VoidVentRecord[] = [];
  let seq = 0;
  const now = Date.now();

  for (let i = 0; i < boundary.counts.length; i++) {
    const count = boundary.counts[i];
    if (count === undefined) {
      continue;
    }
    if (count > 0) {
      vents.push({
        dimensionIdx: i,
        magnitude: Math.round(count),
        sequenceNumber: seq++,
        timestamp: now,
      });
    }
  }

  return buildVoidMerkleTree(vents);
}

/**
 * Build a void Merkle tree from raw dimension counts and round total.
 * Convenience function so consumers don't need to import from void.ts.
 */
export function buildVoidMerkleTreeFromCounts(
  dimensionCounts: readonly number[],
  _totalRounds: number
): VoidMerkleTree {
  const vents: VoidVentRecord[] = [];
  let seq = 0;
  const now = Date.now();
  for (let i = 0; i < dimensionCounts.length; i++) {
    const count = dimensionCounts[i];
    if (count === undefined) {
      continue;
    }
    if (count > 0) {
      vents.push({
        dimensionIdx: i,
        magnitude: Math.round(count),
        sequenceNumber: seq++,
        timestamp: now,
      });
    }
  }
  return buildVoidMerkleTree(vents);
}

/**
 * Build a void Merkle tree by distributing totalRounds across dimensions.
 * Each round rejects one dimension in round-robin order.
 * Convenience for consumers that only know (numDimensions, totalRounds).
 */
export function buildVoidMerkleTreeFromRounds(
  numDimensions: number,
  totalRounds: number
): VoidMerkleTree {
  const counts = new Array(numDimensions).fill(0);
  for (let i = 0; i < totalRounds; i++) {
    counts[i % numDimensions]++;
  }
  return buildVoidMerkleTreeFromCounts(counts, totalRounds);
}

// ============================================================================
// Diagnostics
// ============================================================================

/** Diagnostic summary of a void Merkle tree. */
export interface VoidMerkleDiagnostics {
  /** Root hash -- the void certificate */
  readonly rootHash: number;
  /** Total vent records */
  readonly size: number;
  /** Tree depth (log2 of size, rounded up) */
  readonly depth: number;
  /** Whether the tree is non-empty */
  readonly nonEmpty: boolean;
  /** Total vented magnitude across all records */
  readonly totalVentedMagnitude: number;
  /** Unique dimensions with vent records */
  readonly uniqueDimensions: number;
}

/**
 * Compute diagnostics for a void Merkle tree.
 */
export function computeVoidMerkleDiagnostics(
  tree: VoidMerkleTree
): VoidMerkleDiagnostics {
  const dims = new Set<number>();
  let totalMagnitude = 0;

  for (const vent of tree.vents) {
    dims.add(vent.dimensionIdx);
    totalMagnitude += vent.magnitude;
  }

  return {
    rootHash: tree.rootHash,
    size: tree.size,
    depth: tree.size > 0 ? Math.ceil(Math.log2(tree.size + 1)) : 0,
    nonEmpty: tree.size > 0,
    totalVentedMagnitude: totalMagnitude,
    uniqueDimensions: dims.size,
  };
}
