// @a0n/.ignored_gnosis/src/void-knot-engine.ts
/**
 * The Void — Topological Knot Boss Engine
 *
 * Implements the Snow Crash-inspired ICE combat mechanics.
 * - Bosses are Topological Knots.
 * - Players steal void (hack the knot) over time.
 * - The system fights back with relentless "zombie waves" that grow in intensity.
 * - Group multiplier: damage is split, hacking is parallelized.
 * - Non-zero-sum extraction: the knot doesn't lose void, the player gains it.
 *
 * Forced vulnerability: stealing void triggers the ICE to attack the player's HEX.
 *
 * @module @a0n/gnosis/void-boss-engine
 */

import type {
  KnotEncounterState,
  ObserverState,
  DimensionIndex,
  VoidGameEvent,
  PhyleName,
} from './void-game-types';
import { DIMENSION_TIME_DILATION } from './void-game-types';
import { buleyeanWeight } from './math';

/** Base knot scaling factor per β₁ */
const KNOT_COMPLEXITY_MULTIPLIER = 1000;
/** Base wave damage */
const BASE_WAVE_DAMAGE = 5;

// ── Boss System ─────────────────────────────────────────────────────

/**
 * Initiate a Nexus ICE encounter.
 * The difficulty scales with dimension β₁ and the target Phyle's popularity.
 * Popular phyles have richer DAOs (more void to steal) but harder ICE.
 */
export function initiateKnotEncounter(
  dimension: DimensionIndex,
  targetPhyle: PhyleName,
  phylePopularity: number,
  participants: number
): KnotEncounterState {
  const beta1 = Math.max(1, Math.abs(DIMENSION_TIME_DILATION[dimension] || 1));

  // The knot's total size (void available to steal)
  const maxKnotComplexity = Math.floor(beta1 * KNOT_COMPLEXITY_MULTIPLIER * phylePopularity);

  return {
    knotId: `nexus-${targetPhyle}-${dimension}-${Date.now()}`,
    dimension,
    targetPhyle,
    phylePopularity,
    knotComplexity: maxKnotComplexity,
    maxKnotComplexity,
    systemicEntropy: BASE_WAVE_DAMAGE * beta1,
    ticksElapsed: 0,
    participants: Math.max(0, participants), // Allow 0 to spawn idle
  };
}

export interface KnotTickResult {
  readonly knotState: KnotEncounterState;
  readonly energyRadiated: number;
  readonly voidStolen: number;
  readonly isDefeated: boolean;
  readonly events: VoidGameEvent[];
}

/**
 * Process one tick of the boss encounter (action-based).
 *
 * Forced Vulnerability: Hacking the knot (stealing void) causes the ICE
 * to retaliate. The ICE wave intensity grows linearly over time.
 *
 * Group scaling: Wave damage is divided by the number of participants.
 * Hacking solo is mathematically suicidal at high dimensions.
 */
export function tickKnotEncounter(
  knotState: KnotEncounterState,
  playerState: ObserverState
): KnotTickResult {
  if (knotState.participants <= 0) {
    throw new Error('Cannot tick active combat with 0 participants. Use tickIdleKnotEncounter.');
  }

  const events: VoidGameEvent[] = [];

  // 1. Calculate current wave intensity (grows over time)
  // The longer the encounter takes, the harder the system attacks.
  const timeScaling = 1 + (knotState.ticksElapsed / 10);
  const currentWaveIntensity = Math.floor(knotState.systemicEntropy * timeScaling);

  // 2. Resolve ICE attack vs Player (Forced Vulnerability)
  // God formula determines defense efficiency.
  const effectiveR = playerState.energy;
  const v = currentWaveIntensity;

  const defenseWeight = buleyeanWeight(effectiveR, Math.min(v, effectiveR));
  // The system's chance to hit cleanly vs get deflected by HEX R-pool
  const hitChance = v / (defenseWeight + v + 1);

  // Base damage is the intensity, scaled by group distribution
  let rawDamage = (currentWaveIntensity / knotState.participants);
  if (Math.random() < hitChance) {
    // Critical hit from the wave
    rawDamage *= 1.5;
  }

  const energyRadiated = Math.max(1, Math.floor(rawDamage));

  // 3. Player Steals Void (Hacks the Knot)
  // Steal rate relies on player HEX (stamina) being high.
  // Group dynamics: massively distributed attacks spread the wave damage,
  // but the profit (void extracted per participant) is also divided.
  const rawExtract = (effectiveR / currentWaveIntensity) * 10 * knotState.phylePopularity;
  const voidStolen = Math.floor(Math.max(1, rawExtract / knotState.participants));

  // Update knot state
  const newKnotComplexity = Math.max(0, knotState.knotComplexity - voidStolen);
  const isDefeated = newKnotComplexity === 0;

  const newState: KnotEncounterState = {
    ...knotState,
    knotComplexity: newKnotComplexity,
    ticksElapsed: knotState.ticksElapsed + 1,
  };

  events.push({
    type: 'knot_hacked',
    voidStolen,
    participants: knotState.participants,
    entropySpikeTaken: energyRadiated,
  });

  if (isDefeated) {
    events.push({
      type: 'knot_defeated',
      knotId: knotState.knotId,
      totalTicks: newState.ticksElapsed,
    });
  }

  return {
    knotState: newState,
    energyRadiated,
    voidStolen,
    isDefeated,
    events,
  };
}

/**
 * Allow participants to join or leave an ongoing boss encounter via CRDT state sync.
 * If participants drops to 0, the knot goes idle and begins knowledge spillover.
 */
export function updateBossParticipants(
  knotState: KnotEncounterState,
  newParticipantCount: number
): KnotEncounterState {
  return {
    ...knotState,
    participants: Math.max(0, newParticipantCount),
  };
}

export interface IdleKnotTickResult {
  readonly knotState: KnotEncounterState;
  readonly voidSpilled: number;
}

/**
 * Process a background tick when no observers are actively attacking the knot.
 * Free knowledge spillover: The knot slowly unravels on its own, leaking void into the environment.
 * The ICE defenses also cool down (wave intensity resets).
 */
export function tickIdleKnotEncounter(
  knotState: KnotEncounterState
): IdleKnotTickResult {
  if (knotState.participants > 0) {
    throw new Error('Cannot run idle tick while participants are active');
  }

  // Base wave damage for this dimension
  const baseIntensity = BASE_WAVE_DAMAGE * Math.max(1, Math.abs(DIMENSION_TIME_DILATION[knotState.dimension] || 1));
  
  // Wave intensity cools down slowly when idle
  const newWaveIntensity = Math.max(baseIntensity, knotState.systemicEntropy - 1);
  
  // Time effectively goes backwards for the ICE aggro
  const newTicksElapsed = Math.max(0, knotState.ticksElapsed - 1);

  // Free knowledge spillover: the knot leaks raw void
  // The richer the phyle, the more it leaks.
  const voidSpilled = Math.floor(10 * knotState.phylePopularity);
  const newKnotComplexity = Math.max(0, knotState.knotComplexity - voidSpilled);

  return {
    knotState: {
      ...knotState,
      knotComplexity: newKnotComplexity,
      systemicEntropy: newWaveIntensity,
      ticksElapsed: newTicksElapsed,
    },
    voidSpilled,
  };
}
