// @a0n/.ignored_gnosis/src/void-dimensions.ts
/**
 * Practical void-dimension accounting for software knots.
 *
 * The Lean surface proves that cross-file, temporal, and data-plane gaps are
 * independent contributors to invisible crossings. This module turns those
 * theorem families into a shared tool-facing contract without pretending that a
 * local static detector can recover production-exact counts by itself.
 */

export const VALID_GNOSIS_VOID_DIMENSIONS = [
  'cross_file',
  'temporal',
  'data_plane',
] as const;

export type GnosisVoidDimension = (typeof VALID_GNOSIS_VOID_DIMENSIONS)[number];

export interface GnosisVoidDimensionSignal {
  dimension: GnosisVoidDimension;
  visibleCrossings?: number;
  intermoduleEdges?: number;
  queueDepth?: number;
  dataCrossings?: number;
  exactWitness?: boolean;
}

export interface GnosisVoidDimensionAssessment {
  dimension: GnosisVoidDimension;
  visibleCrossings: number;
  invisibleCrossings: number;
  totalCrossings: number;
  gapIsLowerBound: boolean;
  rationale: string;
  supportingTheorems: readonly string[];
}

function toNat(value: number | undefined | null): number {
  if (!Number.isFinite(value ?? Number.NaN)) {
    return 0;
  }
  return Math.max(0, Math.trunc(value ?? 0));
}

function buildAssessment(
  dimension: GnosisVoidDimension,
  visibleCrossings: number,
  invisibleCrossings: number,
  gapIsLowerBound: boolean,
  rationale: string,
  supportingTheorems: readonly string[]
): GnosisVoidDimensionAssessment {
  return {
    dimension,
    visibleCrossings,
    invisibleCrossings,
    totalCrossings: visibleCrossings + invisibleCrossings,
    gapIsLowerBound,
    rationale,
    supportingTheorems,
  };
}

/**
 * Handles the gnosis assess Void Dimension workflow.
 */
export function assessVoidDimension(
  signal: GnosisVoidDimensionSignal
): GnosisVoidDimensionAssessment {
  const visibleCrossings = toNat(signal.visibleCrossings);
  const gapIsLowerBound = !signal.exactWitness;

  switch (signal.dimension) {
    case 'cross_file': {
      const invisibleCrossings = Math.max(1, toNat(signal.intermoduleEdges));
      return buildAssessment(
        'cross_file',
        visibleCrossings,
        invisibleCrossings,
        gapIsLowerBound,
        'Inter-module ownership edges add global crossings even when the local file looks simple.',
        ['THM-LOCAL-UNKNOT-GLOBAL-KNOT', 'THM-INTERMODULE-GAP-EXACT']
      );
    }
    case 'temporal': {
      const invisibleCrossings = Math.max(1, toNat(signal.queueDepth));
      return buildAssessment(
        'temporal',
        visibleCrossings,
        invisibleCrossings,
        gapIsLowerBound,
        'Load-dependent queueing adds crossings that the static control surface cannot see on its own.',
        ['THM-TEMPORAL-UNKNOT-BECOMES-KNOT', 'THM-OVERLOAD-GAP-EXACT']
      );
    }
    case 'data_plane': {
      const invisibleCrossings = Math.max(1, toNat(signal.dataCrossings));
      return buildAssessment(
        'data_plane',
        visibleCrossings,
        invisibleCrossings,
        gapIsLowerBound,
        'Remote queries and wire hops add data-plane crossings beyond the visible control structure.',
        ['THM-ZERO-CONTROL-POSITIVE-DATA', 'THM-DATA-PLANE-GAP-EXACT']
      );
    }
  }
}
