// @a0n/.ignored_gnosis/src/void-item-catalog.ts
/**
 * The Void — Item Catalog & Drop System
 *
 * Every item in The Void is:
 * 1. Math-based rarity (drop probability = 1/β₁)
 * 2. ZK proof-of-work verified (the item formalizes the proof you were there)
 * 3. Dimension-locked (can only drop in its native dimension)
 * 4. God Formula governed (item effects modify R, v, or w)
 *
 * This is the money maker. Natural value emerges from mathematical
 * improbability. Buyers on Nexus can verify the work was real.
 *
 * @module @a0n/gnosis/void-item-catalog
 */

import type {
  DimensionIndex,
  GnosticDimensionIndex,
  RarityTier,
  GameCoordinate,
  PhyleName,
  ParticleType,
  CosmicFeature,
} from './void-game-types';

import { GNOSTIC } from './math';

// ── Item Categories ─────────────────────────────────────────────────

export type ItemCategory =
  | 'weapon'       // Modify encounter success rate
  | 'armor'        // Reduce HEX cost per action
  | 'consumable'   // One-time use effects
  | 'material'     // Crafting ingredients — DROP THESE
  | 'relic'        // Permanent passive effects (extremely rare)
  | 'cosmetic'     // Visual only (still rare, still valuable)
  | 'key'          // Opens specific dimensional gates
  | 'blueprint';   // Unlocks crafting recipes

export type ItemSlot =
  | 'mainhand'
  | 'offhand'
  | 'helm'
  | 'chest'
  | 'legs'
  | 'boots'
  | 'trinket_1'
  | 'trinket_2'
  | 'none';        // Consumables, materials, blueprints

export type ParticleTypeAffinity =
  | ParticleType
  | 'dps'
  | 'tank'
  | 'fixer'
  | 'range'
  | 'healer';

// ── God Formula Item Effects ────────────────────────────────────────

/**
 * Item effects are expressed as God Formula modifiers.
 * They modify the player's R (budget), v (cost), or w (weight).
 * This keeps balance formally provable.
 */
export interface GodFormulaModifier {
  /** Flat bonus to R (effective HEX pool for encounter resolution) */
  readonly rBonus: number;
  /** Multiplier on encounter difficulty v (< 1 = easier encounters) */
  readonly vMultiplier: number;
  /** Bonus to w (direct weight bonus, stacks additively) */
  readonly wBonus: number;
  /** Reduction to movement HEX cost (additive, min cost remains 1) */
  readonly moveCostReduction: number;
  /** Bonus learning energy per failure */
  readonly learningBonus: number;
  /** Stagnation drain resistance (multiplier, < 1 = slower drain) */
  readonly drainResistance: number;
}

/** No-op modifier — the identity element */
export const IDENTITY_MODIFIER: GodFormulaModifier = {
  rBonus: 0,
  vMultiplier: 1,
  wBonus: 0,
  moveCostReduction: 0,
  learningBonus: 0,
  drainResistance: 1,
} as const;

// ── Item Definition ─────────────────────────────────────────────────

export interface VoidItem {
  /** Unique item template ID */
  readonly itemId: string;
  readonly name: string;
  readonly description: string;
  readonly category: ItemCategory;
  readonly slot: ItemSlot;
  /** Native dimension — can only drop in this dimension */
  readonly nativeDimension: DimensionIndex;
  readonly rarity: RarityTier;
  /** God Formula modifier when equipped */
  readonly modifier: GodFormulaModifier;
  /** Phyle affinity — bonus when used by this phyle */
  readonly phyleAffinity: PhyleName | null;
  /** Archetype affinity — bonus when used by this particleType */
  readonly particleTypeAffinity: ParticleTypeAffinity | null;
  /** Stack size (1 = unique, >1 = stackable materials/consumables) */
  readonly maxStack: number;
  /** Whether this item is tradeable on Nexus */
  readonly tradeable: boolean;
  /** If true, consumed on use */
  readonly consumeOnUse: boolean;
}

// ── Drop Table System ───────────────────────────────────────────────

/**
 * A single entry in a drop table.
 *
 * Drop tables are per-dimension, per-encounter-type.
 * The weight system uses the God Formula:
 *   dropWeight = baseWeight / β₁(dimension)
 *
 * This means:
 * - Common items from Proton (β₁=3): weight/3
 * - Rare items from Void (β₁=21): weight/21
 * - Cosmic items from Pleroma (β₁=55): weight/55
 *
 * The math formalizes the rarity. No arbitrary percentages.
 */
export interface DropTableEntry {
  readonly itemId: string;
  /** Base weight before dimension scaling. Higher = more common within tier. */
  readonly baseWeight: number;
  /** Minimum number of this item to drop */
  readonly minQuantity: number;
  /** Maximum number of this item to drop */
  readonly maxQuantity: number;
  /** Additional condition: require specific cosmic feature on tile */
  readonly requiresFeature?: CosmicFeature;
  /** Additional condition: require minimum failures */
  readonly requiresMinFailures?: number;
  /** Additional condition: require specific phyle */
  readonly requiresPhyle?: PhyleName;
}

export interface DimensionDropTable {
  /** Dimension this table applies to */
  readonly dimension: DimensionIndex;
  /** Common drops (always check) */
  readonly common: readonly DropTableEntry[];
  /** Uncommon drops (check on encounter resolution) */
  readonly uncommon: readonly DropTableEntry[];
  /** Rare drops (check only on failure — failure finds the good stuff) */
  readonly rare: readonly DropTableEntry[];
  /** Boss drop (guaranteed on boss encounter kill) */
  readonly boss: readonly DropTableEntry[];
  /** Secret drops (require specific conditions — discoverable, not documented) */
  readonly secret: readonly DropTableEntry[];
}

// ── The Item Catalog ────────────────────────────────────────────────

// ─── Dimension 0: Bythos (The Point) ────────────────────────────────

const BYTHOS_ITEMS: readonly VoidItem[] = [
  {
    itemId: 'bythos-spark',
    name: 'The First Spark',
    description: 'A fragment of the clinamen itself. The +1 that started everything.',
    category: 'relic',
    slot: 'trinket_1',
    nativeDimension: 0,
    rarity: 'common',
    modifier: { ...IDENTITY_MODIFIER, wBonus: 1 },
    phyleAffinity: null,
    particleTypeAffinity: null,
    maxStack: 1,
    tradeable: true,
    consumeOnUse: false,
  },
  {
    itemId: 'bythos-void-dust',
    name: 'Void Dust',
    description: 'The residue of non-existence. Valuable for crafting.',
    category: 'material',
    slot: 'none',
    nativeDimension: 0,
    rarity: 'common',
    modifier: IDENTITY_MODIFIER,
    phyleAffinity: null,
    particleTypeAffinity: null,
    maxStack: 99,
    tradeable: true,
    consumeOnUse: false,
  },
] as const;

// ─── Dimension 1: Barbelo (The Line) ───────────────────────────────

const BARBELO_ITEMS: readonly VoidItem[] = [
  {
    itemId: 'barbelo-sliver-blade',
    name: 'Sliver Blade',
    description: 'A weapon forged from the first emanation. Cuts in one dimension only.',
    category: 'weapon',
    slot: 'mainhand',
    nativeDimension: 1,
    rarity: 'common',
    modifier: { ...IDENTITY_MODIFIER, rBonus: 3 },
    phyleAffinity: 'nihilist',
    particleTypeAffinity: 'dps',
    maxStack: 1,
    tradeable: true,
    consumeOnUse: false,
  },
  {
    itemId: 'barbelo-thread',
    name: 'Thread of Barbelo',
    description: 'A single thread of the first dimension. Material for line-based crafting.',
    category: 'material',
    slot: 'none',
    nativeDimension: 1,
    rarity: 'common',
    modifier: IDENTITY_MODIFIER,
    phyleAffinity: null,
    particleTypeAffinity: null,
    maxStack: 99,
    tradeable: true,
    consumeOnUse: false,
  },
] as const;

// ─── Dimension 2: Syzygy (The Split) ───────────────────────────────

const SYZYGY_ITEMS: readonly VoidItem[] = [
  {
    itemId: 'syzygy-mirror-shield',
    name: 'Mirror Shield',
    description: 'Reflects half the encounter difficulty back. The antiparallel pair.',
    category: 'armor',
    slot: 'offhand',
    nativeDimension: 2,
    rarity: 'uncommon',
    modifier: { ...IDENTITY_MODIFIER, vMultiplier: 0.85 },
    phyleAffinity: 'philosopher',
    particleTypeAffinity: 'tank',
    maxStack: 1,
    tradeable: true,
    consumeOnUse: false,
  },
  {
    itemId: 'syzygy-fork-stone',
    name: 'Fork Stone',
    description: 'A crystallized fork decision. Material for branching recipes.',
    category: 'material',
    slot: 'none',
    nativeDimension: 2,
    rarity: 'common',
    modifier: IDENTITY_MODIFIER,
    phyleAffinity: null,
    particleTypeAffinity: null,
    maxStack: 99,
    tradeable: true,
    consumeOnUse: false,
  },
  {
    itemId: 'syzygy-learning-potion',
    name: 'Potion of Doubled Failure',
    description: 'Your next failure teaches twice as much. Embrace the vent.',
    category: 'consumable',
    slot: 'none',
    nativeDimension: 2,
    rarity: 'uncommon',
    modifier: { ...IDENTITY_MODIFIER, learningBonus: 5 },
    phyleAffinity: null,
    particleTypeAffinity: null,
    maxStack: 10,
    tradeable: true,
    consumeOnUse: true,
  },
] as const;

// ─── Dimension 3: Proton (The Plane) — Symmetry Zone ────────────────

const PROTON_ITEMS: readonly VoidItem[] = [
  {
    itemId: 'proton-confinement-ring',
    name: 'Confinement Ring',
    description: 'Three quarks bound forever. Reduces movement cost in this symmetry zone.',
    category: 'armor',
    slot: 'trinket_1',
    nativeDimension: 3,
    rarity: 'common',
    modifier: { ...IDENTITY_MODIFIER, moveCostReduction: 0.3 },
    phyleAffinity: 'builder',
    particleTypeAffinity: 'fixer',
    maxStack: 1,
    tradeable: true,
    consumeOnUse: false,
  },
  {
    itemId: 'proton-hex-crystal',
    name: 'Hex Crystal',
    description: 'Crystallized HEX. Restores 25 HEX on use.',
    category: 'consumable',
    slot: 'none',
    nativeDimension: 3,
    rarity: 'common',
    modifier: IDENTITY_MODIFIER,
    phyleAffinity: null,
    particleTypeAffinity: null,
    maxStack: 20,
    tradeable: true,
    consumeOnUse: true,
  },
  {
    itemId: 'proton-gluon-ore',
    name: 'Gluon Ore',
    description: 'The binding force made material. Essential for Plane-tier crafting.',
    category: 'material',
    slot: 'none',
    nativeDimension: 3,
    rarity: 'common',
    modifier: IDENTITY_MODIFIER,
    phyleAffinity: null,
    particleTypeAffinity: null,
    maxStack: 99,
    tradeable: true,
    consumeOnUse: false,
  },
  {
    itemId: 'proton-triad-helm',
    name: 'Triad Helm',
    description: 'Forged from the confinement triad. Three views, one truth.',
    category: 'armor',
    slot: 'helm',
    nativeDimension: 3,
    rarity: 'rare',
    modifier: { ...IDENTITY_MODIFIER, rBonus: 10, vMultiplier: 0.9 },
    phyleAffinity: 'scientist',
    particleTypeAffinity: 'tank',
    maxStack: 1,
    tradeable: true,
    consumeOnUse: false,
  },
] as const;

// ─── Dimension 5: Emanations (Symmetry Zone) ───────────────────────

const EMANATIONS_ITEMS: readonly VoidItem[] = [
  {
    itemId: 'emanations-temporal-bow',
    name: 'Temporal Bow',
    description: 'Fires arrows through time itself. Hits before the target moves.',
    category: 'weapon',
    slot: 'mainhand',
    nativeDimension: 5,
    rarity: 'rare',
    modifier: { ...IDENTITY_MODIFIER, rBonus: 20, vMultiplier: 0.75 },
    phyleAffinity: 'aesthete',
    particleTypeAffinity: 'range',
    maxStack: 1,
    tradeable: true,
    consumeOnUse: false,
  },
  {
    itemId: 'emanations-chronon-shard',
    name: 'Chronon Shard',
    description: 'A frozen moment. Material for time-infused crafting.',
    category: 'material',
    slot: 'none',
    nativeDimension: 5,
    rarity: 'uncommon',
    modifier: IDENTITY_MODIFIER,
    phyleAffinity: null,
    particleTypeAffinity: null,
    maxStack: 50,
    tradeable: true,
    consumeOnUse: false,
  },
  {
    itemId: 'emanations-stagnation-ward',
    name: 'Stagnation Ward',
    description: 'Slows the HEX drain when you stop failing. Buys time to think.',
    category: 'relic',
    slot: 'trinket_2',
    nativeDimension: 5,
    rarity: 'epic',
    modifier: { ...IDENTITY_MODIFIER, drainResistance: 0.5 },
    phyleAffinity: 'mystic',
    particleTypeAffinity: 'healer',
    maxStack: 1,
    tradeable: true,
    consumeOnUse: false,
  },
] as const;

// ─── Dimension 6: Sophia (Non-Euclidean) ────────────────────────────

const SOPHIA_ITEMS: readonly VoidItem[] = [
  {
    itemId: 'sophia-clifford-torus',
    name: 'Clifford Torus',
    description: 'A 4D shape collapsed to your hand. Portals open when you hold it.',
    category: 'relic',
    slot: 'offhand',
    nativeDimension: 6,
    rarity: 'epic',
    modifier: { ...IDENTITY_MODIFIER, moveCostReduction: 0.5, rBonus: 15 },
    phyleAffinity: 'philosopher',
    particleTypeAffinity: null,
    maxStack: 1,
    tradeable: true,
    consumeOnUse: false,
  },
  {
    itemId: 'sophia-topology-ink',
    name: 'Topology Ink',
    description: 'Ink that draws connections between non-adjacent tiles. Crafting material.',
    category: 'material',
    slot: 'none',
    nativeDimension: 6,
    rarity: 'rare',
    modifier: IDENTITY_MODIFIER,
    phyleAffinity: null,
    particleTypeAffinity: null,
    maxStack: 30,
    tradeable: true,
    consumeOnUse: false,
  },
] as const;

// ─── Dimension 8: Void (The Gap) ────────────────────────────────────

const VOID_ITEMS: readonly VoidItem[] = [
  {
    itemId: 'void-gap-walker-boots',
    name: 'Gap Walker Boots',
    description: 'Walk on nothing. Cross void tiles that others cannot.',
    category: 'armor',
    slot: 'boots',
    nativeDimension: 8,
    rarity: 'mythic',
    modifier: { ...IDENTITY_MODIFIER, moveCostReduction: 0.7, drainResistance: 0.3 },
    phyleAffinity: 'hedonist',
    particleTypeAffinity: null,
    maxStack: 1,
    tradeable: true,
    consumeOnUse: false,
  },
  {
    itemId: 'void-web-fragment',
    name: 'Web Fragment',
    description: 'A piece of the gap web itself. Extremely rare crafting material.',
    category: 'material',
    slot: 'none',
    nativeDimension: 8,
    rarity: 'mythic',
    modifier: IDENTITY_MODIFIER,
    phyleAffinity: null,
    particleTypeAffinity: null,
    maxStack: 10,
    tradeable: true,
    consumeOnUse: false,
  },
  {
    itemId: 'void-demiurge-hammer',
    name: 'Demiurge Hammer',
    description: 'The tool of the Demiurge. Gives mass. Generates Landauer heat. Devastating.',
    category: 'weapon',
    slot: 'mainhand',
    nativeDimension: 8,
    rarity: 'mythic',
    modifier: { ...IDENTITY_MODIFIER, rBonus: 50, vMultiplier: 0.6, wBonus: 5 },
    phyleAffinity: 'builder',
    particleTypeAffinity: 'dps',
    maxStack: 1,
    tradeable: true,
    consumeOnUse: false,
  },
] as const;

// ─── Dimension 9: Pleroma (Fullness) ────────────────────────────────

const PLEROMA_ITEMS: readonly VoidItem[] = [
  {
    itemId: 'pleroma-fullness-crown',
    name: 'Crown of Fullness',
    description: 'The golden ratio wrought in metal. Worn by those who mapped the entire void.',
    category: 'armor',
    slot: 'helm',
    nativeDimension: 9,
    rarity: 'cosmic',
    modifier: {
      rBonus: 55, vMultiplier: 0.55, wBonus: 10,
      moveCostReduction: 0.5, learningBonus: 3, drainResistance: 0.2,
    },
    phyleAffinity: null,
    particleTypeAffinity: null,
    maxStack: 1,
    tradeable: true,
    consumeOnUse: false,
  },
  {
    itemId: 'pleroma-phi-blade',
    name: 'Blade of φ',
    description: 'The golden ratio as a weapon. Cuts reality along Fibonacci spirals.',
    category: 'weapon',
    slot: 'mainhand',
    nativeDimension: 9,
    rarity: 'cosmic',
    modifier: { ...IDENTITY_MODIFIER, rBonus: 89, vMultiplier: 0.5, wBonus: 8 },
    phyleAffinity: 'scientist',
    particleTypeAffinity: 'dps',
    maxStack: 1,
    tradeable: true,
    consumeOnUse: false,
  },
  {
    itemId: 'pleroma-essence',
    name: 'Pleroma Essence',
    description: 'Concentrated fullness. The rarest crafting material in existence.',
    category: 'material',
    slot: 'none',
    nativeDimension: 9,
    rarity: 'cosmic',
    modifier: IDENTITY_MODIFIER,
    phyleAffinity: null,
    particleTypeAffinity: null,
    maxStack: 5,
    tradeable: true,
    consumeOnUse: false,
  },
] as const;

// ── Boss Items (One Per Dimension) ──────────────────────────────────

const BOSS_ITEMS: readonly VoidItem[] = [
  {
    itemId: 'boss-barbelo-slivershatter',
    name: 'Slivershatter',
    description: 'Dropped by the Barbelo Guardian. Shatters dimensional boundaries.',
    category: 'weapon',
    slot: 'mainhand',
    nativeDimension: 1,
    rarity: 'rare',
    modifier: { ...IDENTITY_MODIFIER, rBonus: 8, wBonus: 2 },
    phyleAffinity: null,
    particleTypeAffinity: 'dps',
    maxStack: 1,
    tradeable: true,
    consumeOnUse: false,
  },
  {
    itemId: 'boss-proton-quark-mantle',
    name: 'Quark Mantle',
    description: 'Dropped by the Proton Keeper. Three-color armor of the confinement.',
    category: 'armor',
    slot: 'chest',
    nativeDimension: 3,
    rarity: 'epic',
    modifier: { ...IDENTITY_MODIFIER, rBonus: 20, vMultiplier: 0.8, drainResistance: 0.7 },
    phyleAffinity: null,
    particleTypeAffinity: 'tank',
    maxStack: 1,
    tradeable: true,
    consumeOnUse: false,
  },
  {
    itemId: 'boss-sophia-wisdom-tome',
    name: 'Tome of Fallen Wisdom',
    description: 'Dropped by Sophia herself. Each failure teaches 3x with this equipped.',
    category: 'relic',
    slot: 'offhand',
    nativeDimension: 6,
    rarity: 'legendary',
    modifier: { ...IDENTITY_MODIFIER, learningBonus: 10, drainResistance: 0.4 },
    phyleAffinity: 'mystic',
    particleTypeAffinity: 'healer',
    maxStack: 1,
    tradeable: true,
    consumeOnUse: false,
  },
  {
    itemId: 'boss-void-gap-eater',
    name: 'The Gap Eater',
    description: 'Dropped by the Void Archon. Devours distance itself.',
    category: 'weapon',
    slot: 'mainhand',
    nativeDimension: 8,
    rarity: 'mythic',
    modifier: { ...IDENTITY_MODIFIER, rBonus: 55, vMultiplier: 0.5, moveCostReduction: 0.8 },
    phyleAffinity: null,
    particleTypeAffinity: null,
    maxStack: 1,
    tradeable: true,
    consumeOnUse: false,
  },
  {
    itemId: 'boss-pleroma-quine-seed',
    name: 'The Quine Seed',
    description: 'Dropped by the Pleroma itself. Carries the memory of every cycle. The ultimate relic.',
    category: 'relic',
    slot: 'trinket_1',
    nativeDimension: 9,
    rarity: 'cosmic',
    modifier: {
      rBonus: 100, vMultiplier: 0.4, wBonus: 21,
      moveCostReduction: 0.9, learningBonus: 5, drainResistance: 0.1,
    },
    phyleAffinity: null,
    particleTypeAffinity: null,
    maxStack: 1,
    tradeable: true,
    consumeOnUse: false,
  },
] as const;

// ── Master Catalog ──────────────────────────────────────────────────

/** All items in the game, indexed by itemId */
export const ITEM_CATALOG: ReadonlyMap<string, VoidItem> = new Map(
  [
    ...BYTHOS_ITEMS,
    ...BARBELO_ITEMS,
    ...SYZYGY_ITEMS,
    ...PROTON_ITEMS,
    ...EMANATIONS_ITEMS,
    ...SOPHIA_ITEMS,
    ...VOID_ITEMS,
    ...PLEROMA_ITEMS,
    ...BOSS_ITEMS,
  ].map((item) => [item.itemId, item])
);

/** Get all items for a specific dimension */
export function getItemsForDimension(dim: DimensionIndex): VoidItem[] {
  return [...ITEM_CATALOG.values()].filter(
    (item) => item.nativeDimension === dim
  );
}

// ── Drop Tables ─────────────────────────────────────────────────────

/**
 * The β₁-indexed drop table system.
 *
 * Every dimension has its own drop table. Drop probability for each
 * item = (baseWeight / totalWeight) × (1/β₁).
 *
 * This means a Cosmic item from Pleroma (β₁=55) with baseWeight 1 and
 * totalWeight 100 has probability: (1/100) × (1/55) = 1/5500.
 *
 * That's provably rare. And ZK-verified. And tradeable on Nexus.
 */
export const DIMENSION_DROP_TABLES: ReadonlyMap<DimensionIndex, DimensionDropTable> = new Map([
  [0, {
    dimension: 0,
    common: [
      { itemId: 'bythos-void-dust', baseWeight: 50, minQuantity: 1, maxQuantity: 3 },
    ],
    uncommon: [],
    rare: [
      { itemId: 'bythos-spark', baseWeight: 5, minQuantity: 1, maxQuantity: 1 },
    ],
    boss: [],
    secret: [],
  }],
  [1, {
    dimension: 1,
    common: [
      { itemId: 'barbelo-thread', baseWeight: 40, minQuantity: 1, maxQuantity: 5 },
    ],
    uncommon: [
      { itemId: 'barbelo-sliver-blade', baseWeight: 10, minQuantity: 1, maxQuantity: 1 },
    ],
    rare: [],
    boss: [
      { itemId: 'boss-barbelo-slivershatter', baseWeight: 100, minQuantity: 1, maxQuantity: 1 },
    ],
    secret: [],
  }],
  [2, {
    dimension: 2,
    common: [
      { itemId: 'syzygy-fork-stone', baseWeight: 35, minQuantity: 1, maxQuantity: 3 },
    ],
    uncommon: [
      { itemId: 'syzygy-learning-potion', baseWeight: 15, minQuantity: 1, maxQuantity: 2 },
      { itemId: 'syzygy-mirror-shield', baseWeight: 8, minQuantity: 1, maxQuantity: 1 },
    ],
    rare: [],
    boss: [],
    secret: [],
  }],
  [3, {
    dimension: 3,
    common: [
      { itemId: 'proton-gluon-ore', baseWeight: 30, minQuantity: 1, maxQuantity: 5 },
      { itemId: 'proton-hex-crystal', baseWeight: 20, minQuantity: 1, maxQuantity: 2 },
    ],
    uncommon: [
      { itemId: 'proton-confinement-ring', baseWeight: 10, minQuantity: 1, maxQuantity: 1 },
    ],
    rare: [
      { itemId: 'proton-triad-helm', baseWeight: 5, minQuantity: 1, maxQuantity: 1 },
    ],
    boss: [
      { itemId: 'boss-proton-quark-mantle', baseWeight: 100, minQuantity: 1, maxQuantity: 1 },
    ],
    secret: [],
  }],
  [5, {
    dimension: 5,
    common: [
      { itemId: 'emanations-chronon-shard', baseWeight: 25, minQuantity: 1, maxQuantity: 3 },
    ],
    uncommon: [],
    rare: [
      { itemId: 'emanations-temporal-bow', baseWeight: 8, minQuantity: 1, maxQuantity: 1 },
    ],
    boss: [],
    secret: [
      { itemId: 'emanations-stagnation-ward', baseWeight: 3, minQuantity: 1, maxQuantity: 1,
        requiresMinFailures: 100 },
    ],
  }],
  [6, {
    dimension: 6,
    common: [
      { itemId: 'sophia-topology-ink', baseWeight: 20, minQuantity: 1, maxQuantity: 2 },
    ],
    uncommon: [],
    rare: [
      { itemId: 'sophia-clifford-torus', baseWeight: 5, minQuantity: 1, maxQuantity: 1 },
    ],
    boss: [
      { itemId: 'boss-sophia-wisdom-tome', baseWeight: 100, minQuantity: 1, maxQuantity: 1 },
    ],
    secret: [],
  }],
  [8, {
    dimension: 8,
    common: [
      { itemId: 'void-web-fragment', baseWeight: 10, minQuantity: 1, maxQuantity: 1 },
    ],
    uncommon: [],
    rare: [
      { itemId: 'void-gap-walker-boots', baseWeight: 3, minQuantity: 1, maxQuantity: 1 },
      { itemId: 'void-demiurge-hammer', baseWeight: 2, minQuantity: 1, maxQuantity: 1, requiresFeature: 'black_hole' },
    ],
    boss: [
      { itemId: 'boss-void-gap-eater', baseWeight: 100, minQuantity: 1, maxQuantity: 1 },
    ],
    secret: [],
  }],
  [9, {
    dimension: 9,
    common: [],
    uncommon: [],
    rare: [
      { itemId: 'pleroma-essence', baseWeight: 5, minQuantity: 1, maxQuantity: 1 },
      { itemId: 'pleroma-fullness-crown', baseWeight: 2, minQuantity: 1, maxQuantity: 1 },
      { itemId: 'pleroma-phi-blade', baseWeight: 2, minQuantity: 1, maxQuantity: 1 },
    ],
    boss: [
      { itemId: 'boss-pleroma-quine-seed', baseWeight: 100, minQuantity: 1, maxQuantity: 1 },
    ],
    secret: [],
  }],
]);

// ── Drop Resolution ─────────────────────────────────────────────────

export interface ResolvedDrop {
  readonly item: VoidItem;
  readonly quantity: number;
  /** Mathematical improbability of this drop (for display/trading) */
  readonly improbability: number;
}

/**
 * Resolve drops from a dimension's drop table.
 *
 * The math:
 * 1. Select applicable tier (common/uncommon/rare based on outcome)
 * 2. For each entry in the tier: probability = baseWeight/totalWeight × 1/β₁
 * 3. Roll against each entry independently
 * 4. Failed encounters check the RARE table (failure finds the good stuff)
 *
 * @param dimension - Current dimension
 * @param beta1 - β₁ value of this dimension
 * @param isFailure - Whether the encounter was a failure (checks rare table)
 * @param isBoss - Whether this was a boss encounter (checks boss table)
 * @param rng - Seeded PRNG function returning [0, 1)
 * @param playerFailures - Player's total failures (for secret drops)
 * @param tileFeature - Cosmic feature on this tile (for conditional drops)
 * @param playerPhyle - Player's major phyle (for phyle-locked drops)
 */
export function resolveDrops(
  dimension: DimensionIndex,
  beta1: number,
  isFailure: boolean,
  isBoss: boolean,
  rng: () => number,
  playerFailures: number,
  tileFeature: CosmicFeature = 'none',
  playerPhyle?: PhyleName
): ResolvedDrop[] {
  const table = DIMENSION_DROP_TABLES.get(dimension);
  if (!table) return [];
  if (beta1 === 0) beta1 = 1; // Bythos fallback

  const drops: ResolvedDrop[] = [];

  // Always check common
  checkTable(table.common, beta1, rng, drops, playerFailures, tileFeature, playerPhyle);

  // Check uncommon on any encounter
  checkTable(table.uncommon, beta1, rng, drops, playerFailures, tileFeature, playerPhyle);

  // Failure unlocks the RARE table — failure finds the good stuff
  if (isFailure) {
    checkTable(table.rare, beta1, rng, drops, playerFailures, tileFeature, playerPhyle);
    // Secret drops also checked on failure
    checkTable(table.secret, beta1, rng, drops, playerFailures, tileFeature, playerPhyle);
  }

  // Boss: guaranteed drop from boss table
  if (isBoss && table.boss.length > 0) {
    const bossEntry = table.boss[Math.floor(rng() * table.boss.length)];
    if (bossEntry) {
      const item = ITEM_CATALOG.get(bossEntry.itemId);
      if (item) {
        const qty = bossEntry.minQuantity +
          Math.floor(rng() * (bossEntry.maxQuantity - bossEntry.minQuantity + 1));
        drops.push({
          item,
          quantity: qty,
          improbability: beta1 * beta1, // Boss drops: β₁² improbability
        });
      }
    }
  }

  return drops;
}

/** Check a tier's drop table entries against the RNG */
function checkTable(
  entries: readonly DropTableEntry[],
  beta1: number,
  rng: () => number,
  drops: ResolvedDrop[],
  playerFailures: number,
  tileFeature: CosmicFeature,
  playerPhyle?: PhyleName
): void {
  const totalWeight = entries.reduce((sum, e) => sum + e.baseWeight, 0);
  if (totalWeight === 0) return;

  for (const entry of entries) {
    // Check conditions
    if (entry.requiresFeature && entry.requiresFeature !== tileFeature) continue;
    if (entry.requiresMinFailures && playerFailures < entry.requiresMinFailures) continue;
    if (entry.requiresPhyle && entry.requiresPhyle !== playerPhyle) continue;

    // Drop probability = (baseWeight / totalWeight) × (1 / β₁)
    const probability = (entry.baseWeight / totalWeight) / beta1;
    const roll = rng();

    if (roll < probability) {
      const item = ITEM_CATALOG.get(entry.itemId);
      if (item) {
        const qty = entry.minQuantity +
          Math.floor(rng() * (entry.maxQuantity - entry.minQuantity + 1));
        drops.push({
          item,
          quantity: qty,
          improbability: 1 / probability, // How rare this drop is
        });
      }
    }
  }
}

// ── Inventory System ────────────────────────────────────────────────

/** Maximum inventory slots */
export const MAX_INVENTORY_SLOTS = 40;

export interface InventorySlot {
  readonly itemId: string;
  quantity: number;
}

export interface observerEquipment {
  mainhand: VoidItem | null;
  offhand: VoidItem | null;
  helm: VoidItem | null;
  chest: VoidItem | null;
  legs: VoidItem | null;
  boots: VoidItem | null;
  trinket_1: VoidItem | null;
  trinket_2: VoidItem | null;
}

export interface observerInventory {
  readonly slots: InventorySlot[];
  readonly equipment: observerEquipment;
}

/** Create empty inventory */
export function createEmptyInventory(): observerInventory {
  return {
    slots: [],
    equipment: {
      mainhand: null,
      offhand: null,
      helm: null,
      chest: null,
      legs: null,
      boots: null,
      trinket_1: null,
      trinket_2: null,
    },
  };
}

/**
 * Calculate the total God Formula modifier from all equipped items.
 * Modifiers stack additively (R bonus, w bonus, learning bonus)
 * and multiplicatively (v multiplier, drain resistance).
 */
export function calculateEquipmentModifier(
  equipment: observerEquipment
): GodFormulaModifier {
  const equipped = Object.values(equipment).filter(
    (item): item is VoidItem => item !== null
  );

  let rBonus = 0;
  let vMultiplier = 1;
  let wBonus = 0;
  let moveCostReduction = 0;
  let learningBonus = 0;
  let drainResistance = 1;

  for (const item of equipped) {
    rBonus += item.modifier.rBonus;
    vMultiplier *= item.modifier.vMultiplier;
    wBonus += item.modifier.wBonus;
    moveCostReduction = Math.min(0.9, moveCostReduction + item.modifier.moveCostReduction);
    learningBonus += item.modifier.learningBonus;
    drainResistance *= item.modifier.drainResistance;
  }

  return { rBonus, vMultiplier, wBonus, moveCostReduction, learningBonus, drainResistance };
}

/**
 * Calculate a drop's trade value in HEX on Nexus.
 *
 * Value = improbability × β₁_at_drop × (1 + modifier_power)
 *
 * The math determines the price. No arbitrary pricing.
 * The rarer the drop, the more valuable. The higher the dimension
 * it came from, the more valuable. The better the stats, the more valuable.
 */
export function calculateNexusValue(
  item: VoidItem,
  improbability: number,
  beta1AtDrop: number
): number {
  const modPower = item.modifier.rBonus +
    (1 - item.modifier.vMultiplier) * 100 +
    item.modifier.wBonus * 10 +
    item.modifier.moveCostReduction * 50 +
    item.modifier.learningBonus * 20 +
    (1 - item.modifier.drainResistance) * 100;

  return Math.floor(improbability * beta1AtDrop * (1 + modPower / 100));
}
