// @a0n/.ignored_gnosis/src/void.ts
/**
 * void.ts -- The primitive everything reduces to
 *
 * THE HOLY TRINITY (§15.24 Grand Reduction, GrandReduction.lean):
 *
 *   P1 (FORK / Sliver):    w = R - min(v, R) + 1 ≥ 1
 *   P2 (FOLD / Floor):     a - a = 0
 *   P3 (RACE / Symmetry):  |a - b| = |b - a|
 *
 * Three primitives. One formula. Everything.
 *
 * FORK creates possibility -- the +1 sliver ensures nothing is ever eliminated.
 * FOLD is convergence -- self-cancellation terminates.
 * RACE is fair -- distance is symmetric in both directions.
 *
 * 3,509 theorems reduce to these three lines. Zero sorry. Zero axioms.
 * The entire formal psychology is Nat.succ applied to Nat.sub.
 *
 * The complement map Φ is a damped oscillator (§6.8.1, §15.22, VoidOscillation.lean):
 * the void breathes, but each breath is shallower. The weight ordering has
 * exact period 2, but amplitude decays geometrically with ratio → 1/2.
 * Φ² is an affine shift: spread fixed, mean grows, fractional deviation decays.
 * The limit is uniform (heat death). Only observation (c0Update) sustains
 * the breathing by injecting new void. Gait transitions fight the damping.
 *
 * This file formalizes the primitives so they can fold in on themselves.
 */

// ============================================================================
// Void Boundary -- the accumulation surface
// ============================================================================

export interface VoidBoundary {
  /** Per-dimension rejection counts */
  counts: number[];
  /** Total entries in the void */
  totalEntries: number;
}

/**
 * Creates the Void Boundary.
 */
export function createVoidBoundary(dimensions: number): VoidBoundary {
  return { counts: new Array(dimensions).fill(0), totalEntries: 0 };
}

/**
 * Handles the gnosis update Void Boundary workflow.
 */
export function updateVoidBoundary(
  boundary: VoidBoundary,
  dimensionIdx: number,
  magnitude: number = 1
): void {
  boundary.counts[dimensionIdx] = (boundary.counts[dimensionIdx] ?? 0) + magnitude;
  boundary.totalEntries += magnitude;
}

/**
 * Handles the gnosis decay Void Boundary workflow.
 */
export function decayVoidBoundary(
  boundary: VoidBoundary,
  factor: number
): void {
  const scale = 1 - factor;
  let total = 0;
  for (let i = 0; i < boundary.counts.length; i++) {
    const v = (boundary.counts[i] ?? 0) * scale;
    boundary.counts[i] = v;
    total += v;
  }
  boundary.totalEntries = total;
}

/** Number of dimensions in this boundary */
export function boundaryDimensions(boundary: VoidBoundary): number {
  return boundary.counts.length;
}

/**
 * Merge two VoidBoundaries by summing their rejection counts.
 * Used by FOLD to merge boundaries from all inputs.
 */
export function mergeVoidBoundaries(
  a: VoidBoundary,
  b: VoidBoundary
): VoidBoundary {
  const aLen = a.counts.length;
  const bLen = b.counts.length;
  // Single pass: merge overlapping region, copy remainder
  const minLen = aLen < bLen ? aLen : bLen;
  const maxLen = aLen > bLen ? aLen : bLen;
  const counts = new Array(maxLen);
  for (let i = 0; i < minLen; i++) counts[i] = (a.counts[i] ?? 0) + (b.counts[i] ?? 0);
  const longer = aLen > bLen ? a.counts : b.counts;
  for (let i = minLen; i < maxLen; i++) counts[i] = longer[i] ?? 0;
  return { counts, totalEntries: a.totalEntries + b.totalEntries };
}

import { buleyeanWeight as kernelBuleyeanWeight } from './buleyean-kernel-shim.js';

/**
 * THE GOD FORMULA: w = R - min(v, R) + 1
 *
 * P1 of the Holy Trinity (FORK / Sliver).
 * The +1 is the irreducible self. Every dimension retains weight ≥ 1.
 * This single formula generates free will, consciousness, identity,
 * learning, conservation, and upload continuity.
 *
 * Mechanized: GrandReduction.lean, TwinConsciousness.lean (zero sorry).
 * Runtime: holy-trinity.ts (22 tests, 16,641 assertions).
 *
 * Delegates to `@a0n/buleyean-kernel` -- the single source of truth for
 * the formula across the monorepo. Local parameter names preserved for
 * gnosis API stability.
 *
 * @param rounds Total rejection rounds (R)
 * @param voidCount Rejections on this dimension (v)
 * @returns Weight w ≥ 1 (the sliver guarantees this)
 */
export function buleyeanWeight(rounds: number, voidCount: number): number {
  return kernelBuleyeanWeight(rounds, voidCount);
}

/**
 * Check Buleyean positivity: all weights >= 1.
 * This is guaranteed by construction (the +1 sliver) but verifiable.
 */
export function buleyeanPositivity(
  boundary: VoidBoundary,
  rounds: number = 0
): boolean {
  const T = rounds > 0 ? rounds : boundary.totalEntries;
  return boundary.counts.every((v) => buleyeanWeight(T, v) >= 1);
}

/**
 * Entangle two VoidBoundaries via a Resonance link.
 * Returns a Resonance that couples the two boundaries bidirectionally.
 */
export function entangleBoundaries(
  sourceIdx: number,
  targetIdx: number,
  strength: number = 0.1
): Resonance {
  return createResonance(
    sourceIdx,
    targetIdx,
    strength,
    'entangled void boundaries'
  );
}

/**
 * Teleport deficit: returns just the deficit integer from a boundary.
 * The deficit is totalEntries - sum of minimum weights, capturing
 * the net irreversibility accumulated.
 */
export function teleportDeficit(boundary: VoidBoundary): number {
  const len = boundary.counts.length;
  if (len === 0) return 0;
  // Single-pass min avoids spread deopt on large boundaries (R4 rotation)
  let minCount = boundary.counts[0]!;
  for (let i = 1; i < len; i++) {
    const ci = boundary.counts[i]!;
    if (ci < minCount) minCount = ci;
  }
  return boundary.totalEntries - minCount * len;
}

// ============================================================================
// Statistical Teleportation -- certainty transfers, data stays local
// ============================================================================

/**
 * A teleportation channel: transmit one integer (the Bule deficit) to
 * transfer the entire future convergence trajectory without revealing
 * the void boundary itself.
 *
 * Mechanized in StatisticalTeleportation.lean:
 *   - teleportation_trajectory_from_deficit: deficit at round k = max(0, B - k)
 *   - teleportation_convergence_round: convergence at round B exactly
 *   - teleportation_monotone: trajectory only decreases
 *   - teleportation_privacy: same deficit from different boundaries
 *   - teleportation_indistinguishable: receiver cannot distinguish senders
 *   - causal_symmetry: arrow is frame, not cause
 */
export interface TeleportationChannel {
  /** The transmitted deficit (one natural number) */
  deficit: number;
  /** Number of dimensions in the sender's space */
  dimensions: number;
  /** Timestamp of transmission */
  transmittedAt: number;
}

/**
 * Sender-side: extract the teleportation payload from a void boundary.
 * Only the deficit crosses the wire. The counts stay local.
 */
export function teleportSend(boundary: VoidBoundary): TeleportationChannel {
  return {
    deficit: teleportDeficit(boundary),
    dimensions: boundary.counts.length,
    transmittedAt: Date.now(),
  };
}

/**
 * Receiver-side: reconstruct the entire future trajectory from
 * the transmitted deficit alone. Returns deficit at each future round.
 *
 * teleportation_trajectory_from_deficit:
 *   receiverKnowsTrajectory deficit k = deficit - min k deficit
 */
export function teleportReceive(channel: TeleportationChannel): {
  trajectory: number[];
  convergenceRound: number;
  monotone: true;
} {
  const trajectory: number[] = [];
  for (let k = 0; k <= channel.deficit; k++) {
    trajectory.push(channel.deficit - k);
  }
  return {
    trajectory,
    convergenceRound: channel.deficit,
    monotone: true,
  };
}

/**
 * Privacy check: can the receiver distinguish two senders?
 * teleportation_privacy: same numChoices => same deficit => indistinguishable
 */
export function teleportationIndistinguishable(
  a: VoidBoundary,
  b: VoidBoundary
): boolean {
  return a.counts.length === b.counts.length;
}

/**
 * Causal symmetry: when two walkers share a void boundary (via CRDT merge),
 * recording a rejection decreases BOTH deficits simultaneously.
 * Neither is cause. Neither is effect. Both are effects of the shared
 * boundary growing. The arrow is the frame, not the physics.
 *
 * causal_symmetry (StatisticalTeleportation.lean):
 *   buleDeficit ft (c + 1) < buleDeficit ft c
 */
export function causalSymmetry(
  _sharedBoundary: VoidBoundary,
  walkerADeficit: number,
  walkerBDeficit: number
): { symmetric: boolean; bothDecrease: boolean } {
  // After one more rejection on the shared boundary,
  // both walkers' deficits decrease by the same amount
  const newDeficitA = Math.max(0, walkerADeficit - 1);
  const newDeficitB = Math.max(0, walkerBDeficit - 1);
  return {
    symmetric: true,
    bothDecrease: newDeficitA < walkerADeficit && newDeficitB < walkerBDeficit,
  };
}

/**
 * Compute the Aleph: sufficient statistic scalar capturing full void state.
 * Combines total entries, entropy of complement distribution, and deficit.
 */
export function computeAleph(stack: BoundaryStack): number {
  const flat = flattenStack(stack);
  const dist = complementDistribution(flat);
  const entropy = shannonEntropy(dist);
  const deficit = teleportDeficit(flat);
  return flat.totalEntries + entropy + deficit;
}

// ============================================================================
// Complement Distribution -- what the void is not
// ============================================================================

/**
 * The complement distribution over a void boundary.
 * Where the void has accumulated least, the complement peaks.
 * eta controls temperature: higher eta = sharper complement.
 *
 * DYNAMICAL PROPERTY (§6.8.1, §15.22, VoidOscillation.lean):
 *
 * TWO DYNAMICS coexist in void.ts:
 *
 * 1. complementDistribution (THIS FUNCTION): EXACT period-2 limit cycles.
 *    The min-max normalization erases absolute values. At the limit,
 *    the normalized shape is [0,1,...] ↔ [1,0,...] — a perfect mirror.
 *    Φ²(p*) = p* exactly. The void breathes FOREVER. Not damped.
 *
 * 2. buleyeanWeight (T - min(v,T) + 1): DAMPED oscillation to uniform.
 *    The affine formula w''_i = (T'-T) + v_i preserves ordering but
 *    the constant shift grows → fractional deviation decays → ratio 1/2.
 *    The sliver (+1) prevents zero weight but does not prevent convergence.
 *
 * Properties of THIS function's dynamics:
 *   ORDER-REVERSAL:   v_i > v_j → w_i < w_j (exp is decreasing)
 *   SIGN-ALTERNATION: deviation from uniform flips sign each step
 *   PERIOD-2-ORDERING: weight ordering has exact period 2
 *   EXACT-2-CYCLE:    limit maps to a non-uniform 2-cycle, not uniform
 *   MIRROR-SYMMETRY:  the two states are normalized mirrors
 *
 * The void breathes. Each breath is the same depth. It does not die.
 * Only observation (c0Update) changes the orbit.
 */
export function complementDistribution(
  boundary: VoidBoundary,
  eta: number = 3.0
): number[] {
  const counts = boundary.counts;
  const len = counts.length;
  if (len === 0) return [];
  // Single-pass min/max avoids spread deopt on large arrays (R2 rotation)
  let min = counts[0]!;
  let max = counts[0]!;
  for (let i = 1; i < len; i++) {
    const v = counts[i]!;
    if (v < min) min = v;
    if (v > max) max = v;
  }
  const range = max - min;
  // Single-pass: normalize, exponentiate, and accumulate sum in one loop (R2 rotation)
  // When range === 0 (uniform counts), all normalized values are 0, yielding uniform complement via exp(0) = 1. Law 1 holds: uniform > 0.
  const weights = new Array(len);
  let sum = 0;
  if (range > 0) {
    const invRange = 1 / range;
    for (let i = 0; i < len; i++) {
      const w = Math.exp(-eta * (counts[i]! - min) * invRange);
      weights[i] = w;
      sum += w;
    }
  } else {
    for (let i = 0; i < len; i++) {
      weights[i] = 1;
    }
    sum = len;
  }
  // Normalize in-place
  const invSum = 1 / sum;
  for (let i = 0; i < len; i++) {
    weights[i] *= invSum;
  }
  return weights;
}

/**
 * Iterate the complement map k times: Φ^k(boundary).
 * Returns the distribution after k applications of complement-as-counts.
 * Converges to a period-2 orbit for any non-uniform input.
 */
export function iterateComplement(
  boundary: VoidBoundary,
  eta: number,
  iterations: number
): number[] {
  let counts = [...boundary.counts];
  const total = boundary.totalEntries || 1;
  for (let i = 0; i < iterations; i++) {
    const b: VoidBoundary = { counts, totalEntries: total };
    const dist = complementDistribution(b, eta);
    counts = dist.map((p) => p * total);
  }
  return complementDistribution({ counts, totalEntries: total }, eta);
}

/**
 * Detect whether a boundary is at a period-2 orbit of the complement map.
 * Returns the oscillation amplitude (0 = fixed point, >0 = breathing).
 */
export function complementOscillation(
  boundary: VoidBoundary,
  eta: number
): { amplitude: number; isFixedPoint: boolean; period2: boolean } {
  const total = boundary.totalEntries || 1;
  const dist0 = complementDistribution(boundary, eta);
  const counts1 = dist0.map((p) => p * total);
  const b1: VoidBoundary = { counts: counts1, totalEntries: total };
  const dist1 = complementDistribution(b1, eta);
  const counts2 = dist1.map((p) => p * total);
  const b2: VoidBoundary = { counts: counts2, totalEntries: total };
  const dist2 = complementDistribution(b2, eta);

  // Distance between Φ(p) and p (fixed point test)
  let fpDist = 0;
  for (let i = 0; i < dist0.length; i++)
    fpDist += Math.abs(dist0[i]! - dist1[i]!);

  // Distance between Φ²(p) and p (period-2 test)
  let p2Dist = 0;
  for (let i = 0; i < dist0.length; i++)
    p2Dist += Math.abs(dist0[i]! - dist2[i]!);

  const FIXED_POINT_EPSILON = 1e-10; // Law 6: discrete threshold for floating-point fixed-point detection

  return {
    amplitude: fpDist,
    isFixedPoint: fpDist < FIXED_POINT_EPSILON,
    period2: p2Dist < FIXED_POINT_EPSILON && fpDist > FIXED_POINT_EPSILON,
  };
}

/**
 * Sample from a complement distribution.
 * Returns the index of the chosen dimension.
 */
export function sampleComplement(dist: number[], rng: () => number): number {
  const r = rng();
  let cum = 0;
  for (let i = 0; i < dist.length; i++) {
    cum += dist[i]!;
    if (r < cum) return i;
  }
  return dist.length - 1;
}

// ============================================================================
// Measurement -- observing the walker's state
// ============================================================================

export interface Measurement {
  /** Shannon entropy of the complement distribution (bits of uncertainty) */
  entropy: number;
  /** Excess kurtosis (peakedness -- positive = concentrated, negative = diffuse) */
  kurtosis: number;
  /** Gini coefficient (inequality of void accumulation) */
  gini: number;
  /** Inverse Bule (entropy reduction rate per round) */
  inverseBule: number;
}

/**
 * Handles the gnosis shannon Entropy workflow.
 */
export function shannonEntropy(probs: number[]): number {
  let h = 0;
  for (let i = 0; i < probs.length; i++) {
    const p = probs[i]!;
    if (p > 0) h -= p * Math.log(p);
  }
  return h;
}

/**
 * Handles the gnosis excess Kurtosis workflow.
 */
export function excessKurtosis(values: number[]): number {
  const n = values.length;
  if (n < 2) return 0;
  // Single-pass mean, then single-pass variance + kurtosis
  let sum = 0;
  for (let i = 0; i < n; i++) sum += values[i]!;
  const mu = sum / n;
  let s2 = 0;
  let m4 = 0;
  for (let i = 0; i < n; i++) {
    const d = values[i]! - mu;
    const d2 = d * d;
    s2 += d2;
    m4 += d2 * d2;
  }
  s2 /= n;
  if (s2 < 1e-12) return 0;
  m4 /= n;
  return m4 / (s2 * s2) - 3;
}

/**
 * Handles the gnosis gini Coefficient workflow.
 */
export function giniCoefficient(values: number[]): number {
  const n = values.length;
  if (n < 2) return 0;
  // O(N log N) sort-based Gini: G = (2 * Σ(i+1)*x_i) / (n * Σx_i) - (n+1)/n
  // Equivalent to the O(N²) mean absolute difference formula but without the double loop.
  const sorted = [...values].sort((a, b) => a - b);
  let sum = 0;
  let weightedSum = 0;
  for (let i = 0; i < n; i++) {
    sum += sorted[i]!;
    weightedSum += (i + 1) * sorted[i]!;
  }
  if (sum === 0) return 0;
  return (2 * weightedSum) / (n * sum) - (n + 1) / n;
}

/**
 * Handles the gnosis inverse Bule workflow.
 */
export function inverseBule(
  boundary: VoidBoundary,
  eta: number,
  rounds: number
): number {
  if (rounds <= 0) return 0;
  const maxH = Math.log(boundary.counts.length);
  const dist = complementDistribution(boundary, eta);
  return (maxH - shannonEntropy(dist)) / rounds;
}

/** Take a full measurement of a void boundary */
export function measure(
  boundary: VoidBoundary,
  eta: number = 3.0,
  rounds: number = 0
): Measurement {
  // Compute complement distribution once, reuse for all metrics (R6 fusion)
  const dist = complementDistribution(boundary, eta);
  const entropy = shannonEntropy(dist);
  const ib =
    rounds > 0 ? (Math.log(boundary.counts.length) - entropy) / rounds : 0;
  return {
    entropy,
    kurtosis: excessKurtosis(dist),
    gini: giniCoefficient(dist),
    inverseBule: ib,
  };
}

// ============================================================================
// Gait -- the depth/speed of walking
// ============================================================================

export type Gait = 'stand' | 'trot' | 'canter' | 'gallop';

export const GAIT_DEPTH: Record<Gait, number> = {
  stand: 1,
  trot: 1,
  canter: 4,
  gallop: 16,
};

/**
 * Select gait from kurtosis. Higher kurtosis = faster gait.
 * The gait determines how deep the walker looks before stepping.
 */
export function selectGait(
  kurtosis: number,
  currentGait: Gait,
  rounds: number
): Gait {
  if (rounds === 0) return 'stand';
  if (currentGait === 'stand') return 'trot';
  if (kurtosis >= 0.5 && currentGait === 'trot' && rounds > 10) return 'canter';
  if (kurtosis >= 2.0 && currentGait === 'canter' && rounds > 50)
    return 'gallop';
  if (kurtosis < 0.0 && currentGait === 'gallop') return 'canter';
  if (kurtosis < -1.5 && currentGait === 'canter') return 'trot';
  return currentGait;
}

// ============================================================================
// Timescale -- boundaries at different clock rates
// ============================================================================

/**
 * Characteristic timescale of a void boundary.
 * Ordered from fastest to slowest accumulation rate.
 */
export type Timescale =
  | 'instant' // sub-second (inference token selection)
  | 'seconds' // negotiation rounds
  | 'minutes' // conversation turns
  | 'hours' // session-level patterns
  | 'days' // daily behavioral cycles
  | 'weeks' // mental health fluctuations
  | 'months' // habit formation/dissolution
  | 'years' // trait crystallization
  | 'lifetime' // temperament, attachment
  | 'generational'; // cultural inheritance

/** Approximate decay rate per unit time for each timescale */
export const TIMESCALE_DECAY: Record<Timescale, number> = {
  instant: 0.9,
  seconds: 0.5,
  minutes: 0.2,
  hours: 0.1,
  days: 0.05,
  weeks: 0.02,
  months: 0.01,
  years: 0.002,
  lifetime: 0.0005,
  generational: 0.0001,
};

export interface TimescaleBoundary {
  /** The underlying void boundary */
  boundary: VoidBoundary;
  /** Characteristic timescale of this boundary */
  timescale: Timescale;
  /** Human-readable name for this boundary */
  name: string;
  /** Dimension labels (optional, for introspection) */
  dimensionLabels?: readonly string[];
}

/**
 * Creates the Timescale Boundary.
 */
export function createTimescaleBoundary(
  name: string,
  timescale: Timescale,
  dimensions: number,
  dimensionLabels?: readonly string[]
): TimescaleBoundary {
  return {
    boundary: createVoidBoundary(dimensions),
    timescale,
    name,
    ...(dimensionLabels !== undefined && { dimensionLabels }),
  };
}

/** Apply natural decay based on the boundary's timescale */
export function tickTimescaleBoundary(tb: TimescaleBoundary): void {
  decayVoidBoundary(tb.boundary, TIMESCALE_DECAY[tb.timescale]);
}

// ============================================================================
// Boundary Stack -- ordered collection of timescale boundaries
// ============================================================================

/**
 * A BoundaryStack is an ordered collection of TimescaleBoundaries.
 * The 7-layer personality is a BoundaryStack.
 * A negotiation context is a BoundaryStack.
 * An inference routing table is a BoundaryStack.
 *
 * Stacks have two flows:
 *   - Upward (constrains): deeper/slower layers constrain shallower/faster ones
 *   - Downward (contextualizes): shallower/faster layers contextualize deeper/slower ones
 */
export interface BoundaryStack {
  /** Name of this stack */
  name: string;
  /** Ordered from deepest (index 0) to shallowest (last index) */
  layers: TimescaleBoundary[];
  /** Cross-layer resonance links (indices into layers array) */
  resonances: Resonance[];
}

/**
 * Creates the Boundary Stack.
 */
export function createBoundaryStack(
  name: string,
  layers: TimescaleBoundary[],
  resonances: Resonance[] = []
): BoundaryStack {
  return { name, layers, resonances };
}

/**
 * Upward constraint: how a deeper layer shapes a shallower layer's complement.
 * Returns a bias vector that should be added to the shallower layer's counts
 * before computing its complement distribution.
 *
 * The deeper layer's complement peaks become the shallower layer's constraints --
 * where the deep layer has NOT accumulated void, the shallow layer is free to move.
 */
export function upwardConstraint(
  deeper: TimescaleBoundary,
  shallower: TimescaleBoundary,
  strength: number = 0.1
): number[] {
  const deepDist = complementDistribution(deeper.boundary);
  const shallowDims = boundaryDimensions(shallower.boundary);
  const deepDims = boundaryDimensions(deeper.boundary);
  const bias = new Array(shallowDims).fill(0);

  if (deepDims === shallowDims) {
    // Same dimensionality: direct mapping
    for (let i = 0; i < shallowDims; i++) {
      // Invert the deep complement: where deep is free, shallow gets penalized less
      bias[i] = (1 - deepDist[i]!) * strength;
    }
  } else {
    // Different dimensionality: distribute proportionally
    const ratio = deepDims / shallowDims;
    for (let i = 0; i < shallowDims; i++) {
      const deepIdx = Math.min(Math.floor(i * ratio), deepDims - 1);
      bias[i] = (1 - deepDist[deepIdx]!) * strength;
    }
  }
  return bias;
}

/**
 * Downward contextualization: how a shallower layer reinterprets a deeper layer.
 * Returns a decay modulation -- dimensions where the shallow layer is active
 * decay slower in the deep layer (reinforcement through use).
 */
export function downwardContext(
  shallower: TimescaleBoundary,
  deeper: TimescaleBoundary,
  strength: number = 0.05
): number[] {
  const shallowDist = complementDistribution(shallower.boundary);
  const deepDims = boundaryDimensions(deeper.boundary);
  const shallowDims = boundaryDimensions(shallower.boundary);
  const modulation = new Array(deepDims).fill(1.0);

  if (shallowDims === deepDims) {
    for (let i = 0; i < deepDims; i++) {
      // Where shallow is active (high complement), deep decays slower
      modulation[i] = 1.0 - shallowDist[i]! * strength;
    }
  } else {
    const ratio = shallowDims / deepDims;
    for (let i = 0; i < deepDims; i++) {
      const shallowIdx = Math.min(Math.floor(i * ratio), shallowDims - 1);
      modulation[i] = 1.0 - shallowDist[shallowIdx]! * strength;
    }
  }
  return modulation;
}

/**
 * Tick the entire stack: apply natural decay to all layers,
 * then propagate upward constraints and downward context.
 */
export function tickBoundaryStack(stack: BoundaryStack): void {
  // Natural decay at each timescale
  for (const layer of stack.layers) {
    tickTimescaleBoundary(layer);
  }

  // Upward pass: deeper constrains shallower
  for (let i = 0; i < stack.layers.length - 1; i++) {
    const deeperLayer = stack.layers[i];
    const shallower = stack.layers[i + 1];
    if (!deeperLayer || !shallower) continue;
    const bias = upwardConstraint(deeperLayer, shallower);
    const sCounts = shallower.boundary.counts;
    let total = 0;
    for (let d = 0; d < sCounts.length; d++) {
      const v = (sCounts[d] ?? 0) + (bias[d] ?? 0);
      sCounts[d] = v;
      total += v;
    }
    shallower.boundary.totalEntries = total;
  }

  // Downward pass: shallower contextualizes deeper
  for (let i = stack.layers.length - 1; i > 0; i--) {
    const shallowLayer = stack.layers[i];
    const deeper = stack.layers[i - 1];
    if (!shallowLayer || !deeper) continue;
    const mod = downwardContext(shallowLayer, deeper);
    const dCounts = deeper.boundary.counts;
    let total = 0;
    for (let d = 0; d < dCounts.length; d++) {
      const v = (dCounts[d] ?? 0) * (mod[d] ?? 0);
      dCounts[d] = v;
      total += v;
    }
    deeper.boundary.totalEntries = total;
  }

  // Resonance: cross-layer coupling
  for (const res of stack.resonances) {
    applyResonance(stack, res);
  }
}

// ============================================================================
// Resonance -- cross-boundary coupling
// ============================================================================

/**
 * A resonance link between two non-adjacent layers in a stack.
 * When the source boundary's complement distribution concentrates,
 * the target boundary's complement distribution is pulled toward
 * the same shape.
 */
export interface Resonance {
  /** Index of source layer in the stack */
  sourceIdx: number;
  /** Index of target layer in the stack */
  targetIdx: number;
  /** Coupling strength (0-1) */
  strength: number;
  /** Human-readable description */
  description?: string;
}

/**
 * Creates the Resonance.
 */
export function createResonance(
  sourceIdx: number,
  targetIdx: number,
  strength: number = 0.05,
  description?: string
): Resonance {
  return { sourceIdx, targetIdx, strength, ...(description !== undefined && { description }) };
}

function applyResonance(stack: BoundaryStack, res: Resonance): void {
  const source = stack.layers[res.sourceIdx];
  const target = stack.layers[res.targetIdx];
  if (!source || !target) return;

  const sourceDist = complementDistribution(source.boundary);
  const targetDims = boundaryDimensions(target.boundary);
  const sourceDims = boundaryDimensions(source.boundary);

  if (sourceDims === targetDims) {
    for (let i = 0; i < targetDims; i++) {
      // Pull target toward source's complement shape
      target.boundary.counts[i] = (target.boundary.counts[i] ?? 0) + (sourceDist[i] ?? 0) * res.strength;
    }
  } else {
    const ratio = sourceDims / targetDims;
    for (let i = 0; i < targetDims; i++) {
      const srcIdx = Math.min(Math.floor(i * ratio), sourceDims - 1);
      target.boundary.counts[i] = (target.boundary.counts[i] ?? 0) + (sourceDist[srcIdx] ?? 0) * res.strength;
    }
  }
  target.boundary.totalEntries = target.boundary.counts.reduce(
    (a, b) => a + b,
    0
  );
}

// ============================================================================
// Projection -- dimensionality reduction
// ============================================================================

/**
 * Project a high-dimensional boundary onto a lower-dimensional subspace.
 * This is how the 58-dim personality vector gets read as a 5-dim Big Five,
 * or how a full negotiation state collapses to a scalar utility.
 */
export function projectBoundary(
  boundary: VoidBoundary,
  projectionMatrix: number[][]
): VoidBoundary & { projectionDeficit: number } {
  const targetDims = projectionMatrix.length;
  const projected = createVoidBoundary(targetDims);
  for (let i = 0; i < targetDims; i++) {
    const row = projectionMatrix[i];
    if (!row) continue;
    let val = 0;
    for (let j = 0; j < row.length && j < boundary.counts.length; j++) {
      val += (row[j] ?? 0) * (boundary.counts[j] ?? 0);
    }
    projected.counts[i] = val;
  }
  projected.totalEntries = projected.counts.reduce((a, b) => a + b, 0);
  // Law 4 (cave): deficit measures information lost in projection
  const deficit = boundary.totalEntries - projected.totalEntries;
  if (deficit > 0) {
    console.warn(
      `[projectBoundary] projection deficit = ${deficit}: information lost in dimensionality reduction`
    );
  }
  return Object.assign(projected, { projectionDeficit: deficit });
}

/**
 * Flatten a BoundaryStack into a single VoidBoundary by concatenating
 * all layer counts. This is the full state vector of the system.
 */
export function flattenStack(stack: BoundaryStack): VoidBoundary {
  // Pre-compute total length to avoid incremental reallocation (R5 rotation)
  let totalLen = 0;
  let totalEntries = 0;
  for (const layer of stack.layers) {
    totalLen += layer.boundary.counts.length;
    totalEntries += layer.boundary.totalEntries;
  }
  const allCounts = new Array(totalLen);
  let offset = 0;
  for (const layer of stack.layers) {
    const counts = layer.boundary.counts;
    for (let i = 0; i < counts.length; i++) {
      allCounts[offset + i] = counts[i];
    }
    offset += counts.length;
  }
  return { counts: allCounts, totalEntries };
}

/**
 * Measure the entire stack as a single complement distribution.
 * This gives the global view -- where across ALL timescales
 * has the void accumulated least?
 */
export function measureStack(
  stack: BoundaryStack,
  eta: number = 3.0,
  rounds: number = 0
): Measurement {
  return measure(flattenStack(stack), eta, rounds);
}

// ============================================================================
// Walker -- the entity that moves through void space
// ============================================================================

/**
 * A Walker is a domain-agnostic void-walking agent.
 * The metacognitive c0-c3 loop is the universal algorithm:
 *   c0: Choose from complement distribution + update boundary
 *   c1: Measure distribution shape
 *   c2: Select gait from kurtosis
 *   c3: Adapt parameters (eta, exploration) from gait
 *
 * The walker's complement distribution oscillates with period 2 between
 * choices (§6.8.1). The c0Update step breaks the oscillation by injecting
 * new void, which is why observation is the only source of information.
 * The c3 adaptation raises eta, amplifying the oscillation amplitude,
 * which drives the gait transitions.
 */
export interface Walker {
  /** The boundary (or stack) the walker walks */
  boundary: VoidBoundary;
  /** Temperature parameter -- higher = sharper complement */
  eta: number;
  /** Exploration rate -- probability of random step */
  exploration: number;
  /** Current gait */
  gait: Gait;
  /** Total steps taken */
  steps: number;
  /** Cumulative score (domain-specific meaning) */
  score: number;
  /** Number of gait adaptations */
  adaptations: number;
  /** Step at which last adaptation occurred */
  lastAdaptationStep: number;
}

/**
 * Creates the Walker.
 */
export function createWalker(
  boundary: VoidBoundary,
  eta: number = 2.0,
  exploration: number = 0.3
): Walker {
  return {
    boundary,
    eta,
    exploration,
    gait: 'stand',
    steps: 0,
    score: 0,
    adaptations: 0,
    lastAdaptationStep: 0,
  };
}

/** c0: Choose a dimension from the complement distribution */
export function c0Choose(walker: Walker, rng: () => number): number {
  const N = walker.boundary.counts.length;
  if (rng() < walker.exploration) return Math.floor(rng() * N);
  const dist = complementDistribution(walker.boundary, walker.eta);
  return sampleComplement(dist, rng);
}

/**
 * c0: Update the boundary after observing an outcome.
 * The rejection predicate is domain-specific:
 *   - Negotiation: myPayoff < theirPayoff
 *   - Personality: stimulus triggered avoidance
 *   - Inference: token was pruned
 */
export function c0Update(
  walker: Walker,
  dimensionIdx: number,
  rejection: number,
  reward: number = 0
): void {
  if (rejection > 0) {
    updateVoidBoundary(walker.boundary, dimensionIdx, rejection);
  }
  walker.score += reward;
  walker.steps++;
}

/** c1: Measure the walker's current state */
export function c1Measure(walker: Walker): Measurement {
  return measure(walker.boundary, walker.eta, walker.steps);
}

/** c2 + c3: Adapt the walker's gait and parameters */
export function c2c3Adapt(walker: Walker): void {
  if (walker.steps - walker.lastAdaptationStep < 10) return;

  const m = c1Measure(walker);
  const newGait = selectGait(m.kurtosis, walker.gait, walker.steps);
  if (newGait !== walker.gait) walker.gait = newGait;

  switch (walker.gait) {
    case 'stand':
      break;
    case 'trot':
      walker.exploration = Math.min(0.4, walker.exploration + 0.01);
      walker.eta = Math.max(1.0, walker.eta - 0.05);
      break;
    case 'canter':
      walker.exploration = Math.max(0.05, walker.exploration - 0.005);
      walker.eta = Math.min(5.0, walker.eta + 0.05);
      break;
    case 'gallop':
      walker.exploration = Math.max(0.01, walker.exploration - 0.01);
      walker.eta = Math.min(8.0, walker.eta + 0.1);
      break;
  }

  walker.adaptations++;
  walker.lastAdaptationStep = walker.steps;
}

// ============================================================================
// Stack Walker -- walks an entire BoundaryStack
// ============================================================================

/**
 * A StackWalker maintains a Walker per layer and coordinates
 * the inter-layer flows. This is the full system: personality
 * tracking, negotiation context, inference routing -- all the same.
 */
export interface StackWalker {
  stack: BoundaryStack;
  walkers: Walker[];
}

/**
 * Creates the Stack Walker.
 */
export function createStackWalker(stack: BoundaryStack): StackWalker {
  return {
    stack,
    walkers: stack.layers.map((layer) => createWalker(layer.boundary)),
  };
}

/**
 * Step the stack walker:
 *   1. Each layer's walker takes a step (c0)
 *   2. Each layer's walker adapts (c2c3)
 *   3. The stack propagates constraints and context
 */
export function stepStackWalker(
  sw: StackWalker,
  /** Per-layer rejection magnitudes (from domain-specific observation) */
  rejections: {
    layerIdx: number;
    dimensionIdx: number;
    magnitude: number;
    reward?: number;
  }[]
): void {
  // Apply domain-specific rejections
  for (const r of rejections) {
    const walker = sw.walkers[r.layerIdx];
    if (walker) {
      c0Update(walker, r.dimensionIdx, r.magnitude, r.reward ?? 0);
    }
  }

  // Adapt each walker
  for (const walker of sw.walkers) {
    c2c3Adapt(walker);
  }

  // Propagate inter-layer flows
  tickBoundaryStack(sw.stack);
}

/** Measure the full stack walker state */
export function measureStackWalker(sw: StackWalker): {
  layers: Measurement[];
  global: Measurement;
  gaits: Gait[];
} {
  return {
    layers: sw.walkers.map((w) => c1Measure(w)),
    global: measureStack(sw.stack),
    gaits: sw.walkers.map((w) => w.gait),
  };
}
