// @a0n/.ignored_gnosis/src/void-entanglement.ts
/**
 * The Void — Quantum Entanglement & Gravitational Lensing
 *
 * Implements "Guilds/Clans" as Quantum Entanglement Matrices and
 * "Aggro/Tanking" as Deflection via Gravitational Lensing.
 *
 * Observers can entangle their wavefunctions. When entangled,
 * thermodynamic properties (like Knot entropy spikes) are shared across
 * the matrix rather than hitting individuals.
 *
 * Gravitational Lensing mathematically bends incoming entropy toward
 * the observer with the highest Topological Mass (Fermions).
 */

import type { ObserverState, DimensionIndex } from './void-game-types';
import { DIMENSION_TIME_DILATION } from './void-game-types';

export interface EntanglementMatrix {
  readonly matrixId: string;
  readonly constituentDids: readonly string[];
  readonly coherent: boolean; // If false, the entanglement has collapsed
}

/**
 * Entangle multiple observers into a coherent state matrix (a "Guild").
 * Limits to 7 observers per matrix (symmetry constraint).
 */
export function createEntanglementMatrix(dids: string[]): EntanglementMatrix {
  if (dids.length > 7) {
    throw new Error('A matrix cannot sustain coherence with more than 7 constituent observers.');
  }
  return {
    matrixId: `matrix-${Date.now()}-${Math.random().toString(36).substring(2, 7)}`,
    constituentDids: [...new Set(dids)],
    coherent: true
  };
}

/**
 * Calculate the Topological Mass of an Observer.
 * Mass drives gravitational lensing (Aggro).
 * 
 * - Fermions (Tanks) are extremely dense point-particles. Massive baseline multiplier.
 * - Higher dimensions (β₁) are structurally heavier.
 * - Current Energy pool acts as active gravitational distortion.
 */
export function calculateTopologicalMass(observer: ObserverState): number {
  const beta1 = Math.abs(DIMENSION_TIME_DILATION[observer.coordinate.dimension]) || 1;
  const particleMultiplier = observer.particleType === 'fermion' ? 5.0 : 1.0;
  
  // Mass = Energy * β₁ * ParticleMultiplier
  // Fermions bend the space around them dramatically.
  return Math.max(1, observer.energy * beta1 * particleMultiplier);
}

export interface DeflectionResult {
  readonly did: string;
  readonly absorbedEntropy: number;
}

/**
 * Distribute incoming raw entropy (from a Knot) across an Entanglement Matrix.
 * 
 * Instead of splitting damage evenly, entropy behaves like radiation 
 * being bent by gravity. The heaviest topological mass in the matrix 
 * (usually the Fermion) deflects and absorbs the mathematical majority 
 * of the entropy, shielding the lighter Bosons/Anyons.
 * 
 * This is the pure physics mapping of "Threat/Agro".
 */
export function distributeEntropyByMass(
  matrix: EntanglementMatrix,
  observers: ObserverState[], // Must match matrix.constituentDids
  totalRawEntropy: number
): DeflectionResult[] {
  if (!matrix.coherent || observers.length === 0) {
    return [];
  }

  // Calculate total mass of the local matrix
  const masses = observers.map(obs => ({
    did: obs.did,
    mass: calculateTopologicalMass(obs)
  }));

  const totalMass = masses.reduce((sum, item) => sum + item.mass, 0);

  // Distribute entropy proportionally to mass.
  // The heaviest object "attracts" the entropy curve.
  return masses.map(item => {
    const fraction = item.mass / totalMass;
    return {
      did: item.did,
      absorbedEntropy: Math.max(1, Math.floor(totalRawEntropy * fraction))
    };
  });
}
