// @a0n/.ignored_gnosis/src/void-dimensional-mastery.ts
/**
 * The Void — Dimensional Mastery System
 *
 * The skill tree formalizes the dimensional physics. No separate tree needed.
 * Each completed dimension permanently enhances the observer:
 *
 * - Better fast travel (higher β₁ = more spatial compression)
 * - Increased max HEX (health scales with β₁)
 * - Access to rarer drops (higher dimensions = better loot)
 * - Passive modifiers from dimensional understanding
 * - Faster movement through lower dimensions
 *
 * "What do you get at higher dimensionality?" — THIS.
 *
 * The skill tree is not something you choose. It's something you earn
 * by surviving dimensions. The physics rewards you naturally.
 *
 * @module @a0n/gnosis/void-dimensional-mastery
 */

import type {
  DimensionIndex,
  GnosticDimensionIndex,
  ObserverState,
} from './void-game-types';
import { DIMENSION_TIME_DILATION, GNOSTIC_DIMENSIONS } from './void-game-types';
import type { GodFormulaModifier } from './void-item-catalog';
import { IDENTITY_MODIFIER } from './void-item-catalog';
import { GNOSTIC, PHI } from './math';

// ── Dimensional Mastery ─────────────────────────────────────────────

/**
 * What each completed dimension permanently grants.
 *
 * These are not choices — they are physics.
 * Complete dimension N → permanently gain mastery N.
 * The observer formalizes the skill tree, shaped by what it survived.
 */
export interface DimensionalMastery {
  /** Dimension index that was mastered */
  readonly dimension: GnosticDimensionIndex;
  /** Human-readable name */
  readonly name: string;
  /** What this mastery grants */
  readonly description: string;
  /** Permanent modifier to God Formula parameters */
  readonly modifier: GodFormulaModifier;
  /** Max HEX increase (health scaling) */
  readonly energyCapacityBonus: number;
  /** Fast travel compression bonus (additive, stacks) */
  readonly fastTravelBonus: number;
  /** Stagnation grace period extension (picolorenzos) */
  readonly stagnationGrace: number;
}

/**
 * The dimensional mastery table.
 * Each entry is what you permanently gain for completing that dimension.
 *
 * Design philosophy:
 * - Early dimensions: small but meaningful bonuses
 * - Mid dimensions: the power curve bends — you FEEL the difference
 * - Late dimensions: transformative. The observer is fundamentally different.
 * - Pleroma: mastery of everything. The observer transcends.
 *
 * All bonuses scale with β₁ — the math determines the reward.
 */
export const DIMENSIONAL_MASTERIES: ReadonlyMap<GnosticDimensionIndex, DimensionalMastery> = new Map([
  [0, {
    dimension: 0,
    name: 'Point Awareness',
    description: 'You existed in nothing. You understand the clinamen.',
    modifier: { ...IDENTITY_MODIFIER, wBonus: 1 },
    energyCapacityBonus: 0,
    fastTravelBonus: 0,
    stagnationGrace: 0,
  }],
  [1, {
    dimension: 1,
    name: 'Linear Perception',
    description: 'You walked the first line. Movement costs slightly less.',
    modifier: { ...IDENTITY_MODIFIER, moveCostReduction: 0.05 },
    energyCapacityBonus: 10,
    fastTravelBonus: 0.1,
    stagnationGrace: 1,
  }],
  [2, {
    dimension: 2,
    name: 'Bifurcation Sense',
    description: 'You survived the split. You see both paths before choosing.',
    modifier: { ...IDENTITY_MODIFIER, rBonus: 2 },
    energyCapacityBonus: 20,
    fastTravelBonus: 0.15,
    stagnationGrace: 1,
  }],
  [3, {
    dimension: 3,
    name: 'Planar Mastery',
    description: 'The first symmetry zone. The hex grid is home. Encounters cost less.',
    modifier: { ...IDENTITY_MODIFIER, vMultiplier: 0.95, moveCostReduction: 0.1 },
    energyCapacityBonus: 30,
    fastTravelBonus: 0.2,
    stagnationGrace: 2,
  }],
  [4, {
    dimension: 4,
    name: 'Depth Perception',
    description: 'You see beneath surfaces. Hidden tiles sometimes reveal themselves.',
    modifier: { ...IDENTITY_MODIFIER, rBonus: 5, learningBonus: 1 },
    energyCapacityBonus: 50,
    fastTravelBonus: 0.25,
    stagnationGrace: 2,
  }],
  [5, {
    dimension: 5,
    name: 'Temporal Awareness',
    description: 'Time is visible. Stagnation drains slower. You can feel the drain coming.',
    modifier: { ...IDENTITY_MODIFIER, drainResistance: 0.85, learningBonus: 2 },
    energyCapacityBonus: 60,
    fastTravelBonus: 0.3,
    stagnationGrace: 3,
  }],
  [6, {
    dimension: 6,
    name: 'Topological Intuition',
    description: 'Non-Euclidean space bends to your will. Portals cost less. Encounters easier.',
    modifier: { ...IDENTITY_MODIFIER, vMultiplier: 0.85, moveCostReduction: 0.2 },
    energyCapacityBonus: 90,
    fastTravelBonus: 0.4,
    stagnationGrace: 4,
  }],
  [7, {
    dimension: 7,
    name: 'Field Comprehension',
    description: 'You perceive the field. All lower dimensions are transparent.',
    modifier: { ...IDENTITY_MODIFIER, rBonus: 15, vMultiplier: 0.8, learningBonus: 3 },
    energyCapacityBonus: 100,
    fastTravelBonus: 0.5,
    stagnationGrace: 5,
  }],
  [8, {
    dimension: 8,
    name: 'Void Walking',
    description: 'The gap web is yours. You walk where others cannot. The void knows you.',
    modifier: {
      rBonus: 25, vMultiplier: 0.7, wBonus: 3,
      moveCostReduction: 0.3, learningBonus: 4, drainResistance: 0.6,
    },
    energyCapacityBonus: 210,
    fastTravelBonus: 0.7,
    stagnationGrace: 8,
  }],
  [9, {
    dimension: 9,
    name: 'Pleroma Transcendence',
    description: 'Fullness. The golden ratio governs your every step. You are the God Formula.',
    modifier: {
      rBonus: 55, vMultiplier: 0.55, wBonus: 8,
      moveCostReduction: 0.4, learningBonus: 5, drainResistance: 0.4,
    },
    energyCapacityBonus: 550,
    fastTravelBonus: 1.0,
    stagnationGrace: 13,
  }],
]);

// ── Mastery Calculation ─────────────────────────────────────────────

/**
 * Calculate the total mastery modifier for a observer based on
 * their highest completed dimension.
 *
 * All masteries up to and including highestDimension stack.
 * R bonus and w bonus stack additively.
 * v multiplier and drain resistance stack multiplicatively.
 */
export function calculateMasteryModifier(
  highestDimension: number
): GodFormulaModifier {
  let rBonus = 0;
  let vMultiplier = 1;
  let wBonus = 0;
  let moveCostReduction = 0;
  let learningBonus = 0;
  let drainResistance = 1;

  for (let d = 0; d <= Math.min(highestDimension, 9); d++) {
    const mastery = DIMENSIONAL_MASTERIES.get(d as GnosticDimensionIndex);
    if (!mastery) continue;
    rBonus += mastery.modifier.rBonus;
    vMultiplier *= mastery.modifier.vMultiplier;
    wBonus += mastery.modifier.wBonus;
    moveCostReduction = Math.min(0.9, moveCostReduction + mastery.modifier.moveCostReduction);
    learningBonus += mastery.modifier.learningBonus;
    drainResistance *= mastery.modifier.drainResistance;
  }

  return { rBonus, vMultiplier, wBonus, moveCostReduction, learningBonus, drainResistance };
}

/**
 * Calculate the max HEX (health) for a observer based on mastery.
 *
 * Max HEX = BASE + Σ(β₁(d) × 10) for all mastered dimensions d.
 *
 * Health formalizes the gnostic number at each dimension, scaled by 10.
 * The math determines your health pool. No arbitrary numbers.
 *
 * Dim 0 mastery: 100 + 0  = 100  (you're still nothing)
 * Dim 3 mastery: 100 + 60 = 160  (the plane comforts you)
 * Dim 5 mastery: 100 + 170 = 270 (time gives perspective)
 * Dim 9 mastery: 100 + 1120 = 1220 (fullness)
 *
 * The gnostic numbers: 0, 1, 2, 3, 5, 6, 9, 10, 21, 55
 * Sum = 112. × 10 = 1120. + 100 base = 1220 max HEX.
 */
export function calculateEnergyCapacity(highestDimension: number): number {
  const BASE_HEX = 100;
  const HEALTH_PER_BETA = 10;
  let bonus = 0;

  for (let d = 0; d <= Math.min(highestDimension, 9); d++) {
    const dimDef = GNOSTIC_DIMENSIONS.find(
      (dim) => dim.index === d
    );
    if (dimDef) bonus += dimDef.beta1 * HEALTH_PER_BETA;
  }

  return BASE_HEX + bonus;
}

/**
 * Calculate the total stagnation grace period for a observer.
 *
 * Base: 3 picolorenzos
 * Each completed dimension adds grace.
 * Pleroma master: 3 + 0+1+1+2+2+3+4+5+8+13 = 42 picolorenzos
 * (42 — the answer to everything. Not an accident.)
 */
export function calculateStagnationGrace(highestDimension: number): number {
  const BASE_GRACE = 3;
  let bonus = 0;

  for (let d = 0; d <= Math.min(highestDimension, 9); d++) {
    const mastery = DIMENSIONAL_MASTERIES.get(d as GnosticDimensionIndex);
    if (mastery) bonus += mastery.stagnationGrace;
  }

  return BASE_GRACE + bonus;
}

/**
 * Calculate the time (in ticks) an unlogged / inactive player can persist in the void
 * before being vented to the universe.
 * 
 * Scales identically to the thermodynamic time dilation parameter (β₁).
 * Higher dimensions have vastly more structural rigidity, so they evaporate much slower.
 */
export function calculateEvaporationTime(dimension: DimensionIndex): number {
  const BASE_KEEPALIVE_TICKS = 100;
  const beta1 = Math.abs(DIMENSION_TIME_DILATION[dimension]) || 1;
  return BASE_KEEPALIVE_TICKS * beta1;
}

/**
 * Calculate fast travel compression bonus.
 *
 * Base compression: β₁_from / β₁_to
 * Mastery bonus: cumulative from all mastered dimensions.
 *
 * Example: Dim 8→3 base = 21/3 = 7x
 * With full mastery through dim 8: 7x × (1 + 0.7) = 11.9x
 *
 * Teleport gets better with every dimension. Physics.
 */
export function calculateFastTravelCompression(
  fromDimension: DimensionIndex,
  toDimension: DimensionIndex,
  highestDimension: number
): number {
  const fromBeta = Math.abs(DIMENSION_TIME_DILATION[fromDimension]) || 1;
  const toBeta = Math.abs(DIMENSION_TIME_DILATION[toDimension]) || 1;
  const baseCompression = fromBeta / toBeta;

  // Accumulate fast travel bonus from mastered dimensions
  let travelBonus = 0;
  for (let d = 0; d <= Math.min(highestDimension, 9); d++) {
    const mastery = DIMENSIONAL_MASTERIES.get(d as GnosticDimensionIndex);
    if (mastery) travelBonus += mastery.fastTravelBonus;
  }

  return baseCompression * (1 + travelBonus);
}

// ── Optimism Exhaust Logging ────────────────────────────────────────

/**
 * Event types logged to Optimism via dashrelay-app bridge.
 * User doesn't care — but the chain does. Transparent web3.
 *
 * Every significant game action is logged as exhaust.
 * The data is immutable, verifiable, and public.
 * This is how proof-of-work drops are verified on-chain.
 */
export type OptimismExhaustEvent =
  | { type: 'observer_created'; did: string; seed: number; phyle: string; particleType: string }
  | { type: 'dimension_ascended'; did: string; dimension: number; totalFailures: number }
  | { type: 'item_dropped'; did: string; itemId: string; rarity: string; improbability: number; coordinateHash: string }
  | { type: 'item_traded'; fromDid: string; toDid: string; itemId: string; priceHex: number }
  | { type: 'fold_executed'; did: string; cycleCount: number; sliver: number }
  | { type: 'knot_defeated'; did: string; dimension: number; knotId: string }
  | { type: 'mastery_earned'; did: string; dimension: number; masteryName: string };

/**
 * Format an exhaust event for dashrelay-app bridge.
 * Returns the calldata for the Optimism transaction.
 *
 * The bridge is fire-and-forget. Game continues regardless.
 * But the record is permanent.
 */
export function formatExhaustForBridge(event: OptimismExhaustEvent): {
  readonly contractMethod: string;
  readonly args: readonly unknown[];
} {
  return {
    contractMethod: 'logGameExhaust',
    args: [event.type, JSON.stringify(event), Date.now()],
  };
}
