// @a0n/.ignored_gnosis/src/void-combat-engine.ts
/**
 * The Void — Continuous World Combat Engine
 *
 * Implements "Upward Vulnerability" and physical presence combat.
 * Higher-dimensional entities have immense structural rigidity (high β₁).
 * This makes them easy targets for chaotic, high-entropy projections from lower dimensions.
 * Higher-dimensional entities cannot physically attack downward; they must voluntarily
 * "collapse their wavefunction" (drop dimension) to respond to threats.
 */

import type { ObserverState, DimensionIndex, VoidGameEvent, CombatEncounterState } from './void-game-types';
import { CLINAMEN, STARTING_ENERGY } from './void-game-engine';
import { changeDimension } from './void-game-coordinate';
import { buleyeanWeight } from './math';

export interface CombatResult {
  readonly state: ObserverState;
  readonly events: VoidGameEvent[];
  readonly combatState: CombatEncounterState;
  readonly successfulHits: number;
}

export interface PhyleCombatModifier {
  /** Multiplier on the entropy projected during an attack */
  readonly entropyMultiplier: number;
  /** Multiplier on sin accrued */
  readonly sinMultiplier: number;
  /** Efficiency of wavefunction collapse (lower is better/cheaper) */
  readonly collapseCostMultiplier: number;
}

/**
 * Handles the gnosis get Phyle Combat Modifier workflow.
 */
export function getPhyleCombatModifier(phyle: string): PhyleCombatModifier {
  switch (phyle) {
    case 'nihilist':
      return { entropyMultiplier: 1.5, sinMultiplier: 2.0, collapseCostMultiplier: 1.0 };
    case 'builder':
      return { entropyMultiplier: 0.8, sinMultiplier: 1.0, collapseCostMultiplier: 1.2 };
    case 'mystic':
      return { entropyMultiplier: 1.0, sinMultiplier: 0.5, collapseCostMultiplier: 0.1 }; // Extremely cheap collapses
    case 'scientist':
      return { entropyMultiplier: 1.2, sinMultiplier: 1.2, collapseCostMultiplier: 0.8 };
    case 'aesthete':
      return { entropyMultiplier: 0.9, sinMultiplier: 0.8, collapseCostMultiplier: 0.9 };
    case 'hedonist':
      return { entropyMultiplier: 1.1, sinMultiplier: 1.5, collapseCostMultiplier: 1.5 };
    case 'philosopher':
    default:
      return { entropyMultiplier: 1.0, sinMultiplier: 1.0, collapseCostMultiplier: 1.0 };
  }
}



/**
 * Launch an entropy projection at a target.
 * Validates Upward Vulnerability constraints.
 */
export function projectEntropy(
  aggressor: ObserverState,
  targetId: string,
  targetDimension: DimensionIndex,
  targetShardId: string,
  investedEnergy: number
): CombatResult {
  if (investedEnergy < CLINAMEN) {
    throw new Error('Must invest at least the clinamen to project entropy');
  }
  if (aggressor.energy < investedEnergy) {
    throw new Error('Not enough energy to project entropy');
  }

  // Upward Vulnerability Check
  // Cannot attack downward. Must be equal or attacking upward.
  if (aggressor.coordinate.dimension > targetDimension) {
    throw new Error(
      `Structural Block: Cannot project entropy downward from Dimension ${aggressor.coordinate.dimension} to ${targetDimension}. You must collapse your wavefunction to engage.`
    );
  }

  const events: VoidGameEvent[] = [];
  
  // Spend the energy
  const newEnergy = aggressor.energy - investedEnergy;

  // Apply Phyle modifiers
  const aggressorModifiers = getPhyleCombatModifier(aggressor.majorPhyle);

  // Sin is accrued when attacking other observers (selfish entropy)
  // Sin scales relative to the difference in dimension (attacking upward is rebellious, attacking equal is friction)
  const dimensionDelta = Math.max(0, targetDimension - aggressor.coordinate.dimension);
  const sinAccrued = Math.max(1, Math.floor(investedEnergy * (1 - (dimensionDelta * 0.1)) * aggressorModifiers.sinMultiplier));

  // Actual damage payload is scaled
  const finalEntropyPayload = investedEnergy * aggressorModifiers.entropyMultiplier;

  const combatState: CombatEncounterState = {
    encounterId: `combat-${aggressor.did}-${targetId}-${Date.now()}`,
    aggressorDid: aggressor.did,
    targetDid: targetId,
    aggressorDimension: aggressor.coordinate.dimension,
    targetDimension: targetDimension,
    investedEnergy: investedEnergy,
    targetCollapsedWavefunction: false,
    shardId: targetShardId,
  };

  events.push({
    type: 'entropy_projected',
    aggressorDid: aggressor.did,
    targetDid: targetId,
    investedEnergy: investedEnergy,
  });

  if (newEnergy <= 0) {
    events.push({ type: 'death', cause: 'energy_depleted' });
  }

  return {
    state: {
      ...aggressor,
      energy: Math.max(0, newEnergy),
      sin: aggressor.sin + sinAccrued,
    },
    events,
    combatState,
    successfulHits: 1, // Determines base damage payload via UCAN
  };
}

/**
 * Collapse the wavefunction.
 * A higher-dimensional observer intentionally drops to a lower dimension to engage a threat.
 */
export function collapseWavefunction(
  state: ObserverState,
  targetDimension: DimensionIndex
): { state: ObserverState; events: VoidGameEvent[] } {
  if (state.coordinate.dimension <= targetDimension) {
    throw new Error('Wavefunction collapse is only for moving to a lower dimension');
  }

  const events: VoidGameEvent[] = [];

  const targetModifiers = getPhyleCombatModifier(state.majorPhyle);

  // Collapsing a wavefunction releases immense energy into the environment (wasted)
  // The cost is proportional to the dimensional distance modified by the phyle's collapse efficiency
  const dropDistance = state.coordinate.dimension - targetDimension;
  const collapseCost = Math.floor(state.energy * 0.1 * dropDistance * targetModifiers.collapseCostMultiplier);
  const newEnergy = Math.max(CLINAMEN, state.energy - collapseCost);

  const newCoordinate = changeDimension(state.coordinate, targetDimension);

  events.push({
    type: 'wavefunction_collapsed',
    targetDid: state.did,
    fromDimension: state.coordinate.dimension,
    targetDimension: targetDimension,
  });

  return {
    state: {
      ...state,
      energy: newEnergy,
      coordinate: newCoordinate,
    },
    events,
  };
}


/**
 * Metaphysical Realignment (Healing) based on New Thought / Mary Baker Eddy.
 * Healing is not repairing damage; it is affirming the perfect underlying structural integrity.
 * The observer projects Energy not as entropy, but as a structural wave, reducing the target's entropy.
 * Bosons (Healers) excel at this.
 */
export function affirmStructure(
  healer: ObserverState,
  targetId: string,
  targetDimension: DimensionIndex,
  investedEnergy: number
): { state: ObserverState; events: VoidGameEvent[]; transferredEnergy: number } {
  if (investedEnergy < CLINAMEN) {
    throw new Error('Must invest at least the clinamen to affirm structure');
  }
  if (healer.energy < investedEnergy) {
    throw new Error('Not enough energy to affirm structure');
  }
  
  const events: VoidGameEvent[] = [];
  const newEnergy = healer.energy - investedEnergy;

  // Metaphysical Affirmation is "thought based" and "knowledge based" (Mary Baker Eddy).
  // The ability to banish the illusion of entropy is proportional to the healer's Wisdom (Lean/Knowledge).
  const isBoson = healer.particleType === 'boson';
  const baseline = investedEnergy;
  const knowledgeMultiplier = 1 + (healer.wisdom * 0.1); 
  const bosonAmp = isBoson ? 1.5 : 1.0;

  const transferredEnergy = Math.floor(baseline * knowledgeMultiplier * bosonAmp);

  events.push({
    type: 'structure_affirmed',
    healerDid: healer.did,
    targetDid: targetId,
    energyTransferred: transferredEnergy
  });

  return {
    state: {
      ...healer,
      energy: Math.max(0, newEnergy)
    },
    events,
    transferredEnergy
  };
}
