// @a0n/.ignored_gnosis/src/void-sanctuaries.ts
/**
 * The Void — Structural Sanctuaries (Benches)
 *
 * Implements the "Silksong Bench" / Safe Zone mechanic.
 * In a universe driven by thermodynamic entropy, there are specific, perfectly symmetrical
 * coordinates where the mathematics align.
 * 
 * At EVERY Dimension `D`, there is exactly one Home coordinate for every
 * unique permutation of: (MajorPhyle) × (MinorPhyle) × (ParticleType).
 * 
 * Resting at your geometric home restores systemic entropy completely
 * without generating Stagnation penalties. It is a point of utter peace.
 * Sanctuaries also serve as nodes in the inter-dimensional fast-travel network.
 *
 * @module @a0n/gnosis/void-sanctuaries
 */

import type { ObserverState, GameCoordinate, PhyleName, ParticleType, DimensionIndex } from './void-game-types';
import { calculateEnergyCapacity } from './void-dimensional-mastery';

export interface Sanctuary {
  readonly id: string;
  readonly coordinate: GameCoordinate;
  readonly majorPhyle: PhyleName;
  readonly minorPhyle: PhyleName;
  readonly particleType: ParticleType;
}

/**
 * Deterministically calculate the exact (x, y, z) hex coordinate
 * of a Sanctuary for a given combination in a specific dimension.
 * 
 * This uses a deterministic hash so the world generates consistently.
 */
function calculateSanctuaryCoordinate(
  dimension: DimensionIndex,
  majorPhyle: PhyleName,
  minorPhyle: PhyleName,
  particleType: ParticleType
): GameCoordinate {
  // A simple deterministic hash function
  let hash = 0;
  const str = `${dimension}-${majorPhyle}-${minorPhyle}-${particleType}`;
  for (let i = 0; i < str.length; i++) {
    hash = ((hash << 5) - hash) + str.charCodeAt(i);
    hash |= 0;
  }

  // Generate hexQ, hexR, hexS within a 500-hex radius
  const hexQ = (hash % 1000) - 500;
  const hexR = ((hash >>> 3) % 1000) - 500;
  const hexS = -hexQ - hexR; // Hex grid constraint: q + r + s = 0

  return { seed: 0, dimension, hexQ, hexR, hexS, tick: 0 };
}

/**
 * Retrieve the Sanctuary definition for a given Observer's intrinsic physical configuration
 * at a specific dimension.
 */
export function getSanctuaryForObserver(
  observer: ObserverState,
  targetDimension: DimensionIndex
): Sanctuary {
  const coord = calculateSanctuaryCoordinate(
    targetDimension,
    observer.majorPhyle,
    observer.minorPhyle,
    observer.particleType
  );

  return {
    id: `sanctuary-${targetDimension}-${observer.majorPhyle}-${observer.minorPhyle}-${observer.particleType}`,
    coordinate: coord,
    majorPhyle: observer.majorPhyle,
    minorPhyle: observer.minorPhyle,
    particleType: observer.particleType
  };
}

export interface RestResult {
  readonly state: ObserverState;
  readonly energyRestored: number;
}

/**
 * Stop and rest at a sanctuary. 
 * This is the ultimate expression of "Lack of Conflict".
 * Resting restores Energy fully up to the dimensional capacity.
 */
export function restAtSanctuary(observer: ObserverState): RestResult {
  const homeSanctuary = getSanctuaryForObserver(observer, observer.coordinate.dimension);

  if (
    observer.coordinate.hexQ !== homeSanctuary.coordinate.hexQ ||
    observer.coordinate.hexR !== homeSanctuary.coordinate.hexR ||
    observer.coordinate.hexS !== homeSanctuary.coordinate.hexS
  ) {
    throw new Error('You can only rest at your precise topological Home coordinate.');
  }

  const maxEnergy = calculateEnergyCapacity(observer.highestDimension);
  const restoreAmount = Math.max(0, maxEnergy - observer.energy);

  // Resting clears systemic entropy drain accumulation (ploSinceLastFailure)
  return {
    state: {
      ...observer,
      energy: maxEnergy,
      ploSinceLastFailure: 0 
    },
    energyRestored: restoreAmount
  };
}
