// @a0n/.ignored_gnosis/src/void-integration.ts
/**
 * Void Integration -- Voluntary Entry, Settled Ledger, Ki Backpressure
 *
 * TypeScript mirror of VoidIntegration.lean.
 *
 * Three surfaces:
 * 1. Voluntary void entry dominates involuntary vent after breakeven.
 * 2. Settled void ≠ empty void -- serenity has history.
 * 3. Ki channels detect void backpressure before conscious narration.
 */

import type { KiChannels, CoverSpaceState } from './rumination-topology';
import {
  isSerene,
  isRuminating,
  metabolicCostPerRound,
  hasDisharmony,
  isInternalLoop,
} from './rumination-topology';

// ---------------------------------------------------------------------------
// Part I: The Void Ledger
// ---------------------------------------------------------------------------

export interface VoidLedger {
  /** Entries that have been walked, read, integrated */
  processed: number;
  /** Entries generating backpressure -- vented but not yet read */
  unprocessed: number;
  /** Total capacity */
  totalCapacity: number;
}

/** The void is settled: all entries processed, zero backpressure. */
export function isSettled(v: VoidLedger): boolean {
  return v.unprocessed === 0;
}

/** The void is empty: nothing was ever vented. Not serenity -- inexperience. */
export function voidIsEmpty(v: VoidLedger): boolean {
  return v.processed === 0 && v.unprocessed === 0;
}

/** Void backpressure: count of unprocessed entries. */
export function voidBackpressure(v: VoidLedger): number {
  return v.unprocessed;
}

/** Process one void entry: move from unprocessed to processed. */
export function processEntry(v: VoidLedger): VoidLedger {
  if (v.unprocessed < 1) return v;
  return {
    processed: v.processed + 1,
    unprocessed: v.unprocessed - 1,
    totalCapacity: v.totalCapacity,
  };
}

// ---------------------------------------------------------------------------
// Part II: Voluntary vs Involuntary Void Entry
// ---------------------------------------------------------------------------

export type VoidEntryMode = 'voluntary' | 'involuntary';

/** Cost of voluntary entry: one-time spike cost. */
export function voluntaryCost(spikeCost: number): number {
  return spikeCost;
}

/** Cost of involuntary entry: phantom maintenance + trauma. */
export function involuntaryCost(
  phantomPerRound: number,
  rounds: number,
  traumaPenalty: number
): number {
  return rounds * phantomPerRound + traumaPenalty;
}

/** The break-even round: when voluntary becomes cheaper than involuntary. */
export function breakEvenRound(
  spikeCost: number,
  phantomPerRound: number
): number {
  if (phantomPerRound <= 0) return Infinity;
  return Math.ceil(spikeCost / phantomPerRound) + 1;
}

export interface VoluntaryDominanceResult {
  /** Does voluntary dominate at the given round? */
  dominates: boolean;
  /** Voluntary cost */
  vCost: number;
  /** Involuntary cost (no trauma) */
  iCost: number;
  /** Break-even round */
  breakEven: number;
  /** Savings from choosing voluntary */
  savings: number;
}

/** Check if voluntary entry dominates at a given round count. */
export function checkVoluntaryDominance(
  spikeCost: number,
  phantomPerRound: number,
  rounds: number,
  traumaPenalty = 0
): VoluntaryDominanceResult {
  const vCost = voluntaryCost(spikeCost);
  const iCost = involuntaryCost(phantomPerRound, rounds, traumaPenalty);
  return {
    dominates: vCost < iCost,
    vCost,
    iCost,
    breakEven: breakEvenRound(spikeCost, phantomPerRound),
    savings: Math.max(0, iCost - vCost),
  };
}

// ---------------------------------------------------------------------------
// Part III: Ki as Void-Boundary Reader
// ---------------------------------------------------------------------------

/** Void signal: energy + glance leak (internal heat from unprocessed void). */
export function kiVoidSignal(k: KiChannels): number {
  return k.energyLeak + k.glanceLeak;
}

/** Environmental signal: muscle + scent leak (external, not void). */
export function kiEnvironmentalSignal(k: KiChannels): number {
  return k.muscleLeak + k.scentLeak;
}

/** Is the Ki signal dominated by void backpressure? */
export function isVoidDominated(k: KiChannels): boolean {
  return kiVoidSignal(k) > kiEnvironmentalSignal(k) && k.energyLeak >= 1;
}

export type VoidReadingVerdict = 'genuine' | 'illusory';

/** Falsification gate: does the void reading beat 3-way chance? */
export function verifyVoidReading(
  correct: number,
  trials: number
): VoidReadingVerdict {
  return 3 * correct > trials ? 'genuine' : 'illusory';
}

// ---------------------------------------------------------------------------
// Part IV: The Integrated State
// ---------------------------------------------------------------------------

export interface IntegratedState {
  coverSpace: CoverSpaceState;
  voidLedger: VoidLedger;
}

/** Check if a state is fully integrated (serene + settled). */
export function isFullyIntegrated(s: IntegratedState): boolean {
  return isSerene(s.coverSpace) && isSettled(s.voidLedger);
}

// ---------------------------------------------------------------------------
// Part V: Full Void Integration Diagnostic
// ---------------------------------------------------------------------------

export interface VoidIntegrationDiagnostic {
  /** Current void ledger state */
  ledger: VoidLedger;
  /** Is the void settled? */
  settled: boolean;
  /** Is the void empty? */
  empty: boolean;
  /** Current backpressure */
  backpressure: number;
  /** Steps to settle */
  stepsToSettle: number;
  /** Ki signal analysis */
  ki: {
    voidSignal: number;
    environmentalSignal: number;
    voidDominated: boolean;
    disharmony: boolean;
  };
  /** Voluntary vs involuntary analysis */
  entryAnalysis: VoluntaryDominanceResult;
  /** Cover space state */
  coverSpace: {
    serene: boolean;
    ruminating: boolean;
    metabolicCost: number;
    crossingsToResolve: number;
  };
  /** Is the full state integrated? */
  fullyIntegrated: boolean;
  /** Total path to integration (crossings + unprocessed) */
  totalPathToIntegration: number;
}

/**
 * Full void integration diagnostic.
 *
 * Maps Ki channels + cover space + void ledger into a complete analysis:
 * detection, entry mode comparison, settlement path, integration status.
 */
export function diagnoseVoidIntegration(
  cs: CoverSpaceState,
  vl: VoidLedger,
  k: KiChannels,
  spikeCost = 3,
  phantomPerRound = 2,
  coherenceThreshold = 70
): VoidIntegrationDiagnostic {
  const currentRounds = cs.rounds;

  return {
    ledger: vl,
    settled: isSettled(vl),
    empty: voidIsEmpty(vl),
    backpressure: voidBackpressure(vl),
    stepsToSettle: vl.unprocessed,
    ki: {
      voidSignal: kiVoidSignal(k),
      environmentalSignal: kiEnvironmentalSignal(k),
      voidDominated: isVoidDominated(k),
      disharmony: hasDisharmony(k),
    },
    entryAnalysis: checkVoluntaryDominance(
      spikeCost, phantomPerRound, currentRounds
    ),
    coverSpace: {
      serene: isSerene(cs),
      ruminating: isRuminating(cs, coherenceThreshold),
      metabolicCost: metabolicCostPerRound(cs),
      crossingsToResolve: cs.crossings,
    },
    fullyIntegrated: isFullyIntegrated({ coverSpace: cs, voidLedger: vl }),
    totalPathToIntegration: cs.crossings + vl.unprocessed,
  };
}
