// @a0n/.ignored_gnosis/src/wasm-execution.ts
// TODO(betti-wasm phase 4): this top-level node:crypto import (hashArtifact
// fallback ids) keeps GnosisWasmExecutionSurface off the pure edge path; the
// synchronous register() contract makes a lazy import a wider refactor. The
// bridge (wasm-execution-bridge.ts) is already edge-safe for pre-compiled
// Modules via moduleResolver/wasmModule.
import { createHash } from 'node:crypto';
import type { GnosisExecutionAuthContext } from './auth/core';
import type { GnosisCoreCache } from './runtime/core-cache';
import { GnosisRegistry } from './runtime/registry';
import type { GnosisRegistry as SourceGnosisRegistry } from './runtime/registry';
import {
  createWasmBridgeExecutionAst,
  executeWasmWithGnosis,
  registerWasmBridgeHandlers,
  type WasmExecutionAbi,
  type WasmExecutionManifest,
  type WasmModuleResolver,
} from './wasm-execution-bridge';
import {
  attachRequestScopedReactorHandleGrants,
  generateNativeExecutionReactorGrantSigner,
  resolveNativeExecutionAuth,
  summarizeNativeExecutionAuth,
  type NativeExecutionControlHeaders,
  type NativeExecutionAuthSummary,
  type ResolvedNativeExecutionAuth,
  type NativeExecutionReactorGrantSigner,
} from './native-execution-auth';
import {
  canonicalEscInput,
  computeEscCacheKey,
  type EscPureSegmentHint,
  type EscTier,
  type EscWasmSegmentRecord,
} from './esc/edge-side-compute';
import {
  positiveFoldWork,
  wasmEscObservation,
  type CacheQueryResolution,
  type PositiveFoldWork,
  type WasmEscObservation,
} from './esc/cache-observation';

export type { EscPureSegmentHint, EscTier, EscWasmSegmentRecord } from './esc/edge-side-compute';
export type {
  LoadedWasmModule,
  ResolvedWasmExecutionDescriptor,
  WasmModuleResolver,
} from './wasm-execution-bridge';
export type {
  CacheQueryResolution,
  PositiveFoldWork,
  WasmEscObservation,
} from './esc/cache-observation';

export interface GnosisWasmExecutionSurfaceOptions {
  readonly enabled?: boolean;
  readonly coreCache?: GnosisCoreCache | null;
  readonly runtimeBindingHash?: string | null;
  readonly topologyAdmission?: GnosisWasmTopologyAdmission | null;
  readonly reactorGrantAudience?: string;
  /**
   * Optional ESC (edge-side compute) tier. Used only when {@link ExecuteWasmArtifactRequest.esc}
   * is `pure` and no `executionAuth` is present — authenticated runs never teleport.
   */
  readonly escTier?: EscTier | null;
  /**
   * Resolve descriptors to pre-compiled `WebAssembly.Module`s (e.g. Workers
   * `CompiledWasm` imports). Checked before inline base64 / file path — the
   * only module source usable on the edge, where runtime compile is forbidden.
   */
  readonly moduleResolver?: WasmModuleResolver;
}

export interface GnosisWasmTopologyAdmission {
  readonly topologyCoordinates: readonly number[];
  readonly gateCoordinates?: readonly number[];
  readonly decompositionCoordinates?: readonly number[];
  readonly interferePort?: {
    readonly action: string;
    readonly coordinate: number;
    readonly closureCoordinate?: number;
    readonly terminal?: boolean;
  };
  readonly pressureWeightedInterference?: {
    readonly strideTrace?: readonly number[];
    readonly leakagePressureTrace?: readonly number[];
    readonly pressureWeightedStrideTrace?: readonly number[];
    readonly pressureWeightedTotal?: number;
    readonly pressureUnit?: string;
  };
  readonly theoremIds?: readonly string[];
}

export interface WasmExecutionControlRequest {
  readonly method: string;
  readonly path?: string;
  readonly query?: string;
  readonly body?: Uint8Array;
  readonly headers?: Record<string, string>;
}

export interface WasmExecutionControlResponse {
  readonly status: number;
  readonly headers: Record<string, string>;
  readonly body: Uint8Array;
}

interface RegisteredWasmArtifact {
  readonly artifactId: string;
  readonly manifest: WasmExecutionManifest;
  readonly registeredAt: number;
}

export interface RegisterWasmArtifactRequest {
  readonly artifactId?: string;
  readonly wasmPath?: string;
  readonly wasmBase64?: string;
  /**
   * Pre-compiled module (e.g. a deploy-time `CompiledWasm` import). Requires
   * an explicit `artifactId` — Module identity cannot be re-derived by hashing.
   */
  readonly wasmModule?: WebAssembly.Module;
  readonly exportName?: string;
  readonly abi?: WasmExecutionAbi;
  readonly memoryExport?: string;
  readonly allocExport?: string;
  readonly resultLengthExport?: string;
  readonly freeExport?: string;
}

export interface ExecuteWasmArtifactRequest
  extends RegisterWasmArtifactRequest {
  readonly input?: unknown;
  readonly executionAuth?: GnosisExecutionAuthContext | null;
  readonly executionAuthSource?: ResolvedNativeExecutionAuth['source'];
  /**
   * Declared-pure WASM segment hint for ESC teleport. Requires host `escTier` and
   * must not be combined with `executionAuth` (cache disabled when auth is set).
   */
  readonly esc?: EscPureSegmentHint;
}

export interface WasmExecutionResponse {
  readonly artifactId: string;
  readonly exportName: string;
  readonly abi: WasmExecutionAbi;
  readonly payload: unknown;
  readonly logs: string;
  readonly topologyAdmissionFingerprint?: string;
  readonly executionAuth?: NativeExecutionAuthSummary;
  /**
   * When ESC pure mode ran, mirrors `ForkRaceFoldTheorems.CacheObservation` marginal-fold accounting.
   */
  readonly escObservation?: WasmEscObservation;
}

export interface WasmExecutionSurfaceStatus {
  readonly enabled: boolean;
  readonly activeExecutions: number;
  readonly registeredArtifacts: number;
  readonly topologyAdmission?: GnosisWasmTopologyAdmission;
  readonly artifacts: readonly {
    artifactId: string;
    abi: WasmExecutionAbi;
    exportName: string;
    source: 'file' | 'inline' | 'module' | 'resolver';
    registeredAt: number;
  }[];
}

function stableStringify(value: unknown): string {
  if (value === null || value === undefined) {
    return 'null';
  }

  if (typeof value === 'string') {
    return JSON.stringify(value);
  }

  if (typeof value === 'number') {
    return Number.isFinite(value) ? String(value) : 'null';
  }

  if (typeof value === 'bigint') {
    return JSON.stringify(value.toString());
  }

  if (typeof value === 'boolean') {
    return value ? 'true' : 'false';
  }

  if (value instanceof Uint8Array) {
    return `[${Array.from(value).join(',')}]`;
  }

  if (Array.isArray(value)) {
    return `[${value.map((item) => stableStringify(item)).join(',')}]`;
  }

  if (typeof value === 'object') {
    const entries = Object.entries(value)
      .filter(([, entryValue]) => {
        const entryType = typeof entryValue;
        return (
          entryType !== 'undefined' &&
          entryType !== 'function' &&
          entryType !== 'symbol'
        );
      })
      .sort(([leftKey], [rightKey]) => leftKey.localeCompare(rightKey));

    return `{${entries
      .map(
        ([key, entryValue]) =>
          `${JSON.stringify(key)}:${stableStringify(entryValue)}`
      )
      .join(',')}}`;
  }

  return 'null';
}

function hashArtifact(value: unknown): string {
  return createHash('sha256')
    .update(stableStringify(value))
    .digest('hex')
    .slice(0, 24);
}

function stringifyJson(value: unknown): string {
  return JSON.stringify(
    value,
    (_key, candidate) => {
      if (typeof candidate === 'bigint') {
        return candidate.toString();
      }

      if (candidate instanceof Uint8Array) {
        return Array.from(candidate);
      }

      return candidate;
    },
    2
  );
}

function jsonResponse(
  status: number,
  body: unknown
): WasmExecutionControlResponse {
  return {
    status,
    headers: {
      'Content-Type': 'application/json',
      'Cache-Control': 'no-store',
    },
    body: new TextEncoder().encode(stringifyJson(body)),
  };
}

function normalizeRequestPath(request: WasmExecutionControlRequest): {
  readonly pathname: string;
  readonly search: string;
} {
  const path = request.path ?? '/';
  const queryIndex = path.indexOf('?');
  if (queryIndex >= 0) {
    return {
      pathname: path.slice(0, queryIndex) || '/',
      search: path.slice(queryIndex),
    };
  }

  return {
    pathname: path,
    search: request.query ?? '',
  };
}

function parseJsonBody<T>(request: WasmExecutionControlRequest): T {
  if (!request.body || request.body.length === 0) {
    throw new Error('Request body is required.');
  }

  return JSON.parse(
    new TextDecoder('utf-8', { fatal: true }).decode(request.body)
  ) as T;
}

export class GnosisWasmExecutionSurface {
  private readonly enabled: boolean;
  private readonly coreCache: GnosisCoreCache | null;
  private readonly runtimeBindingHash: string | null;
  private readonly topologyAdmission: GnosisWasmTopologyAdmission | null;
  private readonly topologyAdmissionFingerprint: string | null;
  private readonly reactorGrantAudience: string;
  private readonly escTier: EscTier | null;
  private readonly moduleResolver: WasmModuleResolver | null;
  private readonly reactorGrantSignerPromise: Promise<NativeExecutionReactorGrantSigner>;
  private readonly artifacts = new Map<string, RegisteredWasmArtifact>();
  private activeExecutions = 0;

  constructor(options: GnosisWasmExecutionSurfaceOptions = {}) {
    this.enabled = options.enabled ?? false;
    this.coreCache = options.coreCache ?? null;
    this.runtimeBindingHash = options.runtimeBindingHash?.trim() ?? null;
    this.topologyAdmission = options.topologyAdmission ?? null;
    this.topologyAdmissionFingerprint = this.topologyAdmission
      ? hashArtifact({
          kind: 'gnosis-wasm-topology-admission',
          topologyAdmission: this.topologyAdmission,
        })
      : null;
    this.reactorGrantAudience =
      options.reactorGrantAudience ?? 'aeon://gnosis/native-execution';
    this.escTier = options.escTier ?? null;
    this.moduleResolver = options.moduleResolver ?? null;
    this.reactorGrantSignerPromise = generateNativeExecutionReactorGrantSigner(
      this.runtimeBindingHash ?? 'gnosis-wasm-execution'
    );
  }

  registerHandlers(
    registry: GnosisRegistry,
    manifest: WasmExecutionManifest
  ): void {
    registerWasmBridgeHandlers(
      registry as unknown as SourceGnosisRegistry,
      manifest,
      this.moduleResolver ? { moduleResolver: this.moduleResolver } : {}
    );
  }

  register(request: RegisterWasmArtifactRequest): RegisteredWasmArtifact {
    const manifest = this.buildManifest(request);
    if (
      request.artifactId === undefined &&
      (request.wasmModule !== undefined ||
        (manifest.file_path === undefined &&
          manifest.wasm_base64 === undefined))
    ) {
      throw new Error(
        'WASM registration with a pre-compiled wasmModule or a moduleResolver requires an explicit artifactId.'
      );
    }

    const artifactId =
      request.artifactId ??
      hashArtifact({
        file_path: manifest.file_path ?? null,
        wasm_base64: manifest.wasm_base64 ?? null,
        export_name: manifest.export_name,
        abi: manifest.abi,
        memory_export: manifest.memory_export ?? null,
        alloc_export: manifest.alloc_export ?? null,
        result_length_export: manifest.result_length_export ?? null,
        free_export: manifest.free_export ?? null,
      });

    const artifact: RegisteredWasmArtifact = {
      artifactId,
      manifest: { ...manifest, artifact_id: artifactId },
      registeredAt: Date.now(),
    };
    this.artifacts.set(artifactId, artifact);
    return artifact;
  }

  private async attachRequestScopedReactorGrants(
    executionAuth: GnosisExecutionAuthContext | null,
    method: string,
    path: string
  ): Promise<GnosisExecutionAuthContext | null> {
    if (!executionAuth || !this.runtimeBindingHash) {
      return executionAuth;
    }

    return attachRequestScopedReactorHandleGrants({
      executionAuth,
      runtimeBindingHash: this.runtimeBindingHash,
      audience: this.reactorGrantAudience,
      method,
      path,
      signer: await this.reactorGrantSignerPromise,
    });
  }

  async execute(
    request: ExecuteWasmArtifactRequest
  ): Promise<WasmExecutionResponse> {
    const artifact =
      request.artifactId !== undefined
        ? this.artifacts.get(request.artifactId)
        : this.register(request);

    if (!artifact) {
      throw new Error(`Unknown WASM artifact '${request.artifactId}'.`);
    }

    this.activeExecutions += 1;
    try {
      const exportName = artifact.manifest.export_name ?? 'main';
      const abiTag = artifact.manifest.abi ?? 'json-v1';

      const escEnabled =
        request.esc?.mode === 'pure' &&
        this.escTier !== null &&
        request.executionAuth == null;

      let cacheKey: string | undefined;
      let escObservation: WasmEscObservation | undefined;
      let escMissMeta: {
        readonly w: PositiveFoldWork;
        readonly cached: EscWasmSegmentRecord | null;
        readonly signatureMatches: boolean;
        readonly entryValid: boolean;
      } | null = null;

      if (escEnabled && request.esc) {
        const w = positiveFoldWork(request.esc.missFoldWork ?? 1);
        cacheKey = computeEscCacheKey({
          knotSlice: request.esc.knotSlice,
          entryCrossingIndex: request.esc.entryCrossingIndex,
          canonicalInput: canonicalEscInput(request.input),
          artifactFingerprint: artifact.artifactId,
          runtimeBindingHash: this.runtimeBindingHash,
          ...(request.esc.policySalt !== undefined
            ? { policySalt: request.esc.policySalt }
            : {}),
          handlerVersionParts: [
            exportName,
            abiTag,
            ...(this.topologyAdmissionFingerprint
              ? [this.topologyAdmissionFingerprint]
              : []),
          ],
        });

        const cached = await this.escTier!.get(cacheKey);
        const signatureMatches = !!(
          cached &&
          cached.artifactId === artifact.artifactId &&
          cached.exportName === exportName &&
          cached.abi === abiTag
        );
        const entryValid = request.esc.cacheEntryValid !== false;
        escMissMeta = { w, cached, signatureMatches, entryValid };

        if (cached && signatureMatches && entryValid) {
          escObservation = wasmEscObservation({ kind: 'hit' });
          const hitExecutionAuth = summarizeNativeExecutionAuth(
            null,
            request.executionAuthSource ?? 'none'
          );
          const hitResponse: WasmExecutionResponse = {
            artifactId: artifact.artifactId,
            exportName,
            abi: cached.abi as WasmExecutionAbi,
            payload: cached.payload,
            logs: cached.logs,
            ...(this.topologyAdmissionFingerprint
              ? {
                  topologyAdmissionFingerprint:
                    this.topologyAdmissionFingerprint,
                }
              : {}),
            ...(hitExecutionAuth !== undefined
              ? { executionAuth: hitExecutionAuth }
              : {}),
            ...(escObservation !== undefined ? { escObservation } : {}),
          };
          return hitResponse;
        }
      }

      const result = await executeWasmWithGnosis({
        ast: createWasmBridgeExecutionAst(),
        manifest: artifact.manifest,
        input: request.input,
        ...(this.moduleResolver
          ? { moduleResolver: this.moduleResolver }
          : {}),
        ...(this.coreCache
          ? { engineOptions: { coreCache: this.coreCache } }
          : {}),
        ...(request.executionAuth
          ? {
              executeOptions: {
                executionAuth: request.executionAuth,
              },
            }
          : {}),
      });

      if (escEnabled && cacheKey && this.escTier && escMissMeta) {
        const record: EscWasmSegmentRecord = {
          version: 1,
          artifactId: artifact.artifactId,
          exportName,
          abi: abiTag,
          payload: result.payload,
          logs: result.logs,
        };
        await this.escTier.set(cacheKey, record);
        let resolution: CacheQueryResolution;
        if (
          escMissMeta.signatureMatches &&
          escMissMeta.cached !== null &&
          !escMissMeta.entryValid
        ) {
          resolution = { kind: 'stale_miss', w: escMissMeta.w };
        } else {
          resolution = { kind: 'cold_miss', w: escMissMeta.w };
        }
        escObservation = wasmEscObservation(resolution);
      }

      const responseExecutionAuth = summarizeNativeExecutionAuth(
        request.executionAuth ?? null,
        request.executionAuthSource ??
          (request.executionAuth ? 'explicit' : 'none')
      );
      const response: WasmExecutionResponse = {
        artifactId: artifact.artifactId,
        exportName,
        abi: abiTag as WasmExecutionAbi,
        payload: result.payload,
        logs: result.logs,
        ...(this.topologyAdmissionFingerprint
          ? {
              topologyAdmissionFingerprint: this.topologyAdmissionFingerprint,
            }
          : {}),
        ...(responseExecutionAuth !== undefined
          ? { executionAuth: responseExecutionAuth }
          : {}),
        ...(escObservation !== undefined ? { escObservation } : {}),
      };
      return response;
    } finally {
      this.activeExecutions -= 1;
    }
  }

  getStatus(): WasmExecutionSurfaceStatus {
    return {
      enabled: this.enabled,
      activeExecutions: this.activeExecutions,
      registeredArtifacts: this.artifacts.size,
      ...(this.topologyAdmission
        ? { topologyAdmission: this.topologyAdmission }
        : {}),
      artifacts: [...this.artifacts.values()].map((artifact) => ({
        artifactId: artifact.artifactId,
        abi: artifact.manifest.abi ?? 'json-v1',
        exportName: artifact.manifest.export_name ?? 'main',
        source: artifact.manifest.wasm_base64
          ? 'inline'
          : artifact.manifest.wasm_module
            ? 'module'
            : artifact.manifest.file_path
              ? 'file'
              : 'resolver',
        registeredAt: artifact.registeredAt,
      })),
    };
  }

  async handleRequest(
    request: WasmExecutionControlRequest
  ): Promise<WasmExecutionControlResponse | null> {
    if (!this.enabled) {
      return null;
    }

    const { pathname } = normalizeRequestPath(request);

    if (pathname === '/.aeon/wasm/status' && request.method === 'GET') {
      return jsonResponse(200, this.getStatus());
    }

    if (pathname === '/.aeon/wasm/register' && request.method === 'POST') {
      try {
        const body = parseJsonBody<RegisterWasmArtifactRequest>(request);
        const artifact = this.register(body);
        return jsonResponse(200, {
          artifactId: artifact.artifactId,
          abi: artifact.manifest.abi ?? 'json-v1',
          exportName: artifact.manifest.export_name ?? 'main',
        });
      } catch (error) {
        return jsonResponse(400, {
          error: error instanceof Error ? error.message : String(error),
        });
      }
    }

    if (pathname === '/.aeon/wasm/execute' && request.method === 'POST') {
      try {
        const body = parseJsonBody<ExecuteWasmArtifactRequest>(request);
        const authOptions: {
          headers?: NativeExecutionControlHeaders;
          explicitExecutionAuth?: GnosisExecutionAuthContext | null;
          expectedReactorGrantAudience?: string;
        } = {};
        if (request.headers !== undefined) {
          authOptions.headers = request.headers;
        }
        if (body.executionAuth !== undefined) {
          authOptions.explicitExecutionAuth = body.executionAuth;
        }
        if (this.reactorGrantAudience !== undefined) {
          authOptions.expectedReactorGrantAudience = this.reactorGrantAudience;
        }
        const executionAuth = await resolveNativeExecutionAuth(authOptions);
        const requestExecutionAuth =
          await this.attachRequestScopedReactorGrants(
            executionAuth.executionAuth,
            request.method,
            pathname
          );
        const result = await this.execute({
          ...body,
          ...(requestExecutionAuth
            ? {
                executionAuth: requestExecutionAuth,
                executionAuthSource: executionAuth.source,
              }
            : {}),
        });
        return jsonResponse(200, result);
      } catch (error) {
        return jsonResponse(400, {
          error: error instanceof Error ? error.message : String(error),
        });
      }
    }

    return null;
  }

  private buildManifest(
    request: RegisterWasmArtifactRequest
  ): WasmExecutionManifest {
    if (
      (request.wasmPath === undefined || request.wasmPath.length === 0) &&
      (request.wasmBase64 === undefined || request.wasmBase64.length === 0) &&
      request.wasmModule === undefined &&
      this.moduleResolver === null
    ) {
      throw new Error(
        'WASM registration requires wasmPath, wasmBase64, wasmModule, or a surface-level moduleResolver.'
      );
    }

    return {
      ...(request.wasmPath ? { file_path: request.wasmPath } : {}),
      ...(request.wasmBase64 ? { wasm_base64: request.wasmBase64 } : {}),
      ...(request.wasmModule ? { wasm_module: request.wasmModule } : {}),
      ...(request.exportName ? { export_name: request.exportName } : {}),
      ...(request.abi ? { abi: request.abi } : {}),
      ...(request.memoryExport ? { memory_export: request.memoryExport } : {}),
      ...(request.allocExport ? { alloc_export: request.allocExport } : {}),
      ...(request.resultLengthExport
        ? { result_length_export: request.resultLengthExport }
        : {}),
      ...(request.freeExport ? { free_export: request.freeExport } : {}),
      ...(request.wasmPath ? { resolve_relative_to: request.wasmPath } : {}),
      node_execution_plans: [
        {
          node_id: 'wasm_entry',
          kind: 'entry',
        },
        {
          node_id: 'wasm_call',
          kind: 'call',
        },
        {
          node_id: 'wasm_return',
          kind: 'return',
        },
      ],
    };
  }
}
