// @a0n/.ignored_gnosis/src/wasm-substrate.ts
/**
 * wasm-substrate.ts -- Universal WASM acceleration for gnosis core operations
 *
 * Auto-detects WASM availability and transparently accelerates gnosis
 * primitives using the buleyean-probability and fano-omega crates.
 * Falls back to pure-TS implementations when WASM is unavailable.
 *
 * Accelerated operations:
 *   - godWeight      (buleyean-probability: god_weight)
 *   - buleyeanDist   (buleyean-probability: buleyean_distribution)
 *   - buleNumber     (buleyean-probability: bule_number)
 *   - shannonEntropy (buleyean-probability: shannon_entropy)
 *   - fanoCollide    (fano-omega: triton_collide)
 *   - omegaSolve     (fano-omega: omega_solve)
 *   + batch variants for all of the above
 *
 * WASM modules are loaded lazily on first call. If loading fails
 * (missing .wasm file, no WebAssembly support, test environment),
 * all functions silently fall back to the pure-TS path.
 *
 * @see wasm-modules/buleyean-probability/
 * @see wasm-modules/fano-omega/
 * @see open-source/gnosis/src/void.ts (pure-TS fallbacks)
 * @see open-source/gnosis/src/holy-trinity.ts (godFormula)
 */

// ─── WASM module state ──────────────────────────────────────────────────────

let buleyeanWasm: any = null;
let fanoWasm: any = null;
let initAttempted = false;
let initPromise: Promise<void> | null = null;

/**
 * Current acceleration status.
 */
export interface SubstrateStatus {
  /** Whether WASM init has been attempted */
  attempted: boolean;
  /** Whether buleyean-probability WASM is loaded */
  buleyean: boolean;
  /** Whether fano-omega WASM is loaded */
  fano: boolean;
}

/**
 * Handles the gnosis substrate Status workflow.
 */
export function substrateStatus(): SubstrateStatus {
  return {
    attempted: initAttempted,
    buleyean: buleyeanWasm !== null,
    fano: fanoWasm !== null,
  };
}

// ─── WASM initialization ────────────────────────────────────────────────────

/**
 * Initialize WASM modules. Safe to call multiple times; only loads once.
 * Returns silently on failure -- callers never need to handle errors.
 */
export function initWasmSubstrate(): Promise<void> {
  if (initPromise) return initPromise;
  initPromise = _doInit();
  return initPromise;
}

function loadBuleyeanWasm(): Promise<void> {
  return import(
    '../../../wasm-modules/buleyean-probability/pkg/buleyean_probability.js'
  ).then(
    (mod) => {
      const ready =
        typeof mod.default === 'function'
          ? Promise.resolve(mod.default())
          : Promise.resolve();
      return ready.then(
        () => {
          buleyeanWasm = mod;
        },
        () => {
          buleyeanWasm = null;
        }
      );
    },
    () => {
      buleyeanWasm = null;
    }
  );
}

function loadFanoWasm(): Promise<void> {
  return import('../../../wasm-modules/fano-omega/pkg/fano_omega.js').then(
    (mod) => {
      const ready =
        typeof mod.default === 'function'
          ? Promise.resolve(mod.default())
          : Promise.resolve();
      return ready.then(
        () => {
          fanoWasm = mod;
        },
        () => {
          fanoWasm = null;
        }
      );
    },
    () => {
      fanoWasm = null;
    }
  );
}

function _doInit(): Promise<void> {
  initAttempted = true;
  return loadBuleyeanWasm().then(loadFanoWasm, loadFanoWasm);
}

/**
 * Force-reset WASM state. Useful for testing fallback paths.
 */
export function resetWasmSubstrate(): void {
  buleyeanWasm = null;
  fanoWasm = null;
  initAttempted = false;
  initPromise = null;
}

// ─── Pure-TS fallback implementations ───────────────────────────────────────

/**
 * God Formula (pure TS): w = R - min(v, R) + 1
 */
function _tsGodWeight(R: number, v: number): number {
  return R - Math.min(v, R) + 1;
}

/**
 * Buleyean distribution (pure TS): normalize god-weights to a probability
 * distribution. Sum = 1, all entries > 0.
 */
function _tsBuleyeanDistribution(
  rejectionCounts: number[],
  round: number
): number[] {
  const n = rejectionCounts.length;
  if (n === 0) return [];
  const weights: number[] = new Array(n);
  let sum = 0;
  for (let i = 0; i < n; i++) {
    const w = _tsGodWeight(round, rejectionCounts[i]!);
    weights[i] = w;
    sum += w;
  }
  if (sum === 0) {
    // Uniform fallback (should not happen since god_weight >= 1)
    const u = 1 / n;
    return new Array(n).fill(u);
  }
  const inv = 1 / sum;
  for (let i = 0; i < n; i++) {
    weights[i]! *= inv;
  }
  return weights;
}

/**
 * Bule number (pure TS): max(counts) - min(counts). 0 = perfect convergence.
 */
function _tsBuleNumber(counts: number[]): number {
  if (counts.length === 0) return 0;
  let min = counts[0]!;
  let max = counts[0]!;
  for (let i = 1; i < counts.length; i++) {
    const v = counts[i]!;
    if (v < min) min = v;
    if (v > max) max = v;
  }
  return max - min;
}

/**
 * Shannon entropy (pure TS): -sum(p * log2(p)) for p > 0.
 */
function _tsShannonEntropy(distribution: number[]): number {
  let h = 0;
  for (let i = 0; i < distribution.length; i++) {
    const p = distribution[i]!;
    if (p > 0) h -= p * Math.log2(p);
  }
  return h;
}

/**
 * Fano collide / triton collide (pure TS): XOR on Fano plane points (mod 7).
 * Points are 0-6 in GF(8)\{0} mapped through the Fano XOR table.
 */
function _tsFanoCollide(a: number, b: number): number {
  // XOR in GF(8) restricted to Fano plane points 0-6
  // The Fano plane has 7 points. triton_collide is raw XOR.
  return a ^ b;
}

/**
 * Omega solve (pure TS): transition a 9-attribute Fano state vector.
 * Each attribute is XOR-collided with its force.
 */
function _tsOmegaSolve(
  current: Uint8Array,
  forces: Uint8Array
): Uint8Array {
  const n = Math.min(current.length, forces.length);
  const result = new Uint8Array(n);
  for (let i = 0; i < n; i++) {
    result[i] = current[i]! ^ forces[i]!;
  }
  return result;
}

// ─── Accelerated exports ────────────────────────────────────────────────────

/**
 * God weight: w = R - min(v, R) + 1
 *
 * Uses WASM `god_weight` when available, pure TS otherwise.
 */
export function godWeight(R: number, v: number): number {
  if (buleyeanWasm) {
    return buleyeanWasm.god_weight(R, v);
  }
  return _tsGodWeight(R, v);
}

/**
 * Buleyean distribution: normalized god-weights. Sum = 1.
 *
 * Uses WASM `buleyean_distribution` when available.
 *
 * @param rejectionCounts Per-agent rejection counts
 * @param round Current round number (R)
 * @returns Probability distribution (sum = 1, all entries > 0)
 */
export function buleyeanDistribution(
  rejectionCounts: number[],
  round: number
): number[] {
  if (buleyeanWasm) {
    const counts = new Uint32Array(rejectionCounts);
    const result = buleyeanWasm.buleyean_distribution(
      counts,
      round,
      rejectionCounts.length
    );
    return Array.from(result);
  }
  return _tsBuleyeanDistribution(rejectionCounts, round);
}

/**
 * Bule number: max(counts) - min(counts). 0 = convergence.
 *
 * Uses WASM `bule_number` when available.
 */
export function buleNumber(counts: number[]): number {
  if (buleyeanWasm) {
    const arr = new Uint32Array(counts);
    return buleyeanWasm.bule_number(arr, counts.length);
  }
  return _tsBuleNumber(counts);
}

/**
 * Shannon entropy: -sum(p * log2(p)). Returns bits.
 *
 * Uses WASM `shannon_entropy` when available.
 */
export function shannonEntropy(distribution: number[]): number {
  if (buleyeanWasm) {
    const arr = new Float64Array(distribution);
    return buleyeanWasm.shannon_entropy(arr, distribution.length);
  }
  return _tsShannonEntropy(distribution);
}

/**
 * Fano collide: XOR collision on Fano plane points.
 *
 * Uses WASM `triton_collide` when available.
 */
export function fanoCollide(a: number, b: number): number {
  if (fanoWasm) {
    return fanoWasm.triton_collide(a, b);
  }
  return _tsFanoCollide(a, b);
}

/**
 * Omega solve: transition a 9-attribute Fano state vector.
 * Input: two 9-byte arrays (current + forces). Output: 9-byte next state.
 *
 * Uses WASM `omega_solve` when available.
 */
export function omegaSolve(
  current: Uint8Array,
  forces: Uint8Array
): Uint8Array {
  if (fanoWasm) {
    return fanoWasm.omega_solve(current, forces);
  }
  return _tsOmegaSolve(current, forces);
}

// ─── Batch variants ─────────────────────────────────────────────────────────

/**
 * Batch god weight: compute w for N agents at once.
 *
 * Uses WASM `god_weight_batch` when available. Returns plain number array.
 */
export function godWeightBatch(
  rounds: number[],
  rejections: number[]
): number[] {
  const count = Math.min(rounds.length, rejections.length);
  if (buleyeanWasm) {
    const r = new Uint32Array(rounds);
    const v = new Uint32Array(rejections);
    const result = buleyeanWasm.god_weight_batch(r, v, count);
    return Array.from(result);
  }
  const out: number[] = new Array(count);
  for (let i = 0; i < count; i++) {
    out[i] = _tsGodWeight(rounds[i]!, rejections[i]!);
  }
  return out;
}

/**
 * Batch buleyean distribution: compute distributions for N groups.
 *
 * Uses WASM `buleyean_distribution_batch` when available.
 *
 * @param allCounts Flat array of rejection counts (groups concatenated)
 * @param agentsPerGroup Number of agents per group
 * @param groupCount Number of groups
 * @returns Flat probability array (groups concatenated)
 */
export function buleyeanDistributionBatch(
  allCounts: number[],
  agentsPerGroup: number,
  groupCount: number
): number[] {
  if (buleyeanWasm) {
    const counts = new Uint32Array(allCounts);
    const result = buleyeanWasm.buleyean_distribution_batch(
      counts,
      agentsPerGroup,
      groupCount
    );
    return Array.from(result);
  }
  // TS fallback: process each group individually
  const out: number[] = [];
  for (let g = 0; g < groupCount; g++) {
    const offset = g * agentsPerGroup;
    const groupCounts = allCounts.slice(offset, offset + agentsPerGroup);
    // Use totalEntries as round approximation (sum of counts in group)
    const round = groupCounts.reduce((a, b) => a + b, 0);
    const dist = _tsBuleyeanDistribution(groupCounts, round);
    out.push(...dist);
  }
  return out;
}

/**
 * Batch shannon entropy: compute entropy for N groups of distributions.
 *
 * Uses WASM `entropy_batch` when available.
 *
 * @param distributions Flat probability array (groups concatenated)
 * @param agentsPerGroup Entries per group
 * @param groupCount Number of groups
 * @returns One entropy value per group
 */
export function shannonEntropyBatch(
  distributions: number[],
  agentsPerGroup: number,
  groupCount: number
): number[] {
  if (buleyeanWasm) {
    const arr = Float64Array.from(distributions);
    const result = buleyeanWasm.entropy_batch(arr, agentsPerGroup, groupCount);
    return Array.from(result);
  }
  const out: number[] = new Array(groupCount);
  for (let g = 0; g < groupCount; g++) {
    const offset = g * agentsPerGroup;
    const group = distributions.slice(offset, offset + agentsPerGroup);
    out[g] = _tsShannonEntropy(group);
  }
  return out;
}

/**
 * Batch bule number: compute bule for N groups.
 *
 * Uses WASM `bule_batch` when available.
 */
export function buleNumberBatch(
  allCounts: number[],
  agentsPerGroup: number,
  groupCount: number
): number[] {
  if (buleyeanWasm) {
    const counts = new Uint32Array(allCounts);
    const result = buleyeanWasm.bule_batch(counts, agentsPerGroup, groupCount);
    return Array.from(result);
  }
  const out: number[] = new Array(groupCount);
  for (let g = 0; g < groupCount; g++) {
    const offset = g * agentsPerGroup;
    const group = allCounts.slice(offset, offset + agentsPerGroup);
    out[g] = _tsBuleNumber(group);
  }
  return out;
}

/**
 * Batch fano collide: XOR-collide N pairs.
 */
export function fanoCollideBatch(
  aValues: number[],
  bValues: number[]
): number[] {
  const count = Math.min(aValues.length, bValues.length);
  const out: number[] = new Array(count);
  if (fanoWasm) {
    for (let i = 0; i < count; i++) {
      out[i] = fanoWasm.triton_collide(aValues[i]!, bValues[i]!);
    }
  } else {
    for (let i = 0; i < count; i++) {
      out[i] = _tsFanoCollide(aValues[i]!, bValues[i]!);
    }
  }
  return out;
}

/**
 * Batch omega solve: transition N state vectors in one call.
 *
 * Uses WASM `omega_solve_batch` when available.
 *
 * @param current Flat array of N * 9 current-state bytes
 * @param forces Flat array of N * 9 force bytes
 * @param count Number of state vectors
 * @returns Flat array of N * 9 next-state bytes
 */
export function omegaSolveBatch(
  current: Uint8Array,
  forces: Uint8Array,
  count: number
): Uint8Array {
  if (fanoWasm) {
    return fanoWasm.omega_solve_batch(current, forces, count);
  }
  // TS fallback: process each 9-byte vector
  const stride = 9;
  const result = new Uint8Array(count * stride);
  for (let i = 0; i < count; i++) {
    const off = i * stride;
    const c = current.subarray(off, off + stride);
    const f = forces.subarray(off, off + stride);
    const r = _tsOmegaSolve(c, f);
    result.set(r, off);
  }
  return result;
}
