// @a0n/.ignored_gnosis/src/wormhole-topology.ts
/**
 * Wormhole Topology — connecting spacelike-separated events and rewiring
 * the causal graph.
 *
 * A wormhole (Einstein-Rosen bridge) connects two mouths at
 * spacelike-separated events, creating a shortcut through spacetime.
 * This module computes how inserting such a bridge rewires the causal
 * structure: pairs that were VENT (spacelike, acausal) may become
 * PROCESS (timelike, causal) once the wormhole provides a traversable
 * path.
 *
 * @packageDocumentation
 */

import type {
  SpacetimeEvent,
  CausalTopologyPacket,
  CausalEdge,
  CausalEdgeType,
} from './causal-topology';
import {
  spacetimeInterval,
  createCausalTopology,
  lightConeRegion,
} from './causal-topology';
import { LIGHT_PICOLORENZO_LY } from './math';

// ─── Types ─────────────────────────────────────────────────────────

/**
 * A wormhole connection between two spacetime events.
 */
export interface WormholeConnection {
  /** Event ID for mouth A. */
  readonly mouthA: string;
  /** Event ID for mouth B. */
  readonly mouthB: string;
  /** Time to traverse the wormhole in picolorenzos (0 = instant). */
  readonly traversalTimePLo: number;
  /** Optional human-readable label. */
  readonly label?: string;
}

/**
 * An edge that was reclassified by the wormhole insertion.
 */
export interface RewiredEdge {
  readonly sourceId: string;
  readonly targetId: string;
  readonly originalEdgeType: CausalEdgeType;
  readonly newEdgeType: CausalEdgeType;
}

/**
 * Result of inserting a wormhole into a causal topology.
 */
export interface WormholeTopologyResult {
  readonly source: 'GnosisWormholeTopology';
  readonly connection: WormholeConnection;
  readonly originalTopology: CausalTopologyPacket;
  readonly modifiedTopology: CausalTopologyPacket;
  readonly rewiredEdges: readonly RewiredEdge[];
  readonly newlyCausalPairs: readonly {
    readonly eventA: string;
    readonly eventB: string;
  }[];
  readonly originalCausalConnectivity: number;
  readonly modifiedCausalConnectivity: number;
  readonly connectivityGain: number;
}

// ─── Internal Helpers ──────────────────────────────────────────────

/**
 * Build a one-hop causal reachability map from the edge list.
 * Returns a map: eventId -> Set of event IDs reachable via a single
 * timelike or lightlike edge where eventId is the source.
 */
function buildForwardReach(
  edges: readonly CausalEdge[]
): Map<string, Set<string>> {
  const reach = new Map<string, Set<string>>();
  for (const edge of edges) {
    if (edge.edgeType === 'VENT') continue;
    let set = reach.get(edge.sourceId);
    if (!set) {
      set = new Set();
      reach.set(edge.sourceId, set);
    }
    set.add(edge.targetId);
  }
  return reach;
}

/**
 * Build reverse one-hop causal reachability: eventId -> Set of event IDs
 * that can reach eventId via a single causal edge (eventId is target).
 */
function buildReverseReach(
  edges: readonly CausalEdge[]
): Map<string, Set<string>> {
  const reach = new Map<string, Set<string>>();
  for (const edge of edges) {
    if (edge.edgeType === 'VENT') continue;
    let set = reach.get(edge.targetId);
    if (!set) {
      set = new Set();
      reach.set(edge.targetId, set);
    }
    set.add(edge.sourceId);
  }
  return reach;
}

/**
 * Compute transitive closure of forward reachability from a starting event.
 * BFS over the forward reach map.
 */
function transitiveForwardReach(
  startId: string,
  forward: Map<string, Set<string>>
): Set<string> {
  const visited = new Set<string>();
  const queue = [startId];
  while (queue.length > 0) {
    const current = queue.pop()!;
    const nexts = forward.get(current);
    if (!nexts) continue;
    for (const next of nexts) {
      if (!visited.has(next)) {
        visited.add(next);
        queue.push(next);
      }
    }
  }
  return visited;
}

/**
 * Compute transitive closure of reverse reachability from a starting event.
 */
function transitiveReverseReach(
  startId: string,
  reverse: Map<string, Set<string>>
): Set<string> {
  const visited = new Set<string>();
  const queue = [startId];
  while (queue.length > 0) {
    const current = queue.pop()!;
    const prevs = reverse.get(current);
    if (!prevs) continue;
    for (const prev of prevs) {
      if (!visited.has(prev)) {
        visited.add(prev);
        queue.push(prev);
      }
    }
  }
  return visited;
}

// ─── Core Functions ────────────────────────────────────────────────

/**
 * Insert a wormhole into a causal topology and compute the rewired graph.
 *
 * Algorithm:
 * 1. Find the events for mouthA and mouthB.
 * 2. Build transitive forward and reverse reachability from existing
 *    causal (timelike + lightlike) edges.
 * 3. The wormhole adds a bidirectional shortcut: any event X that can
 *    causally reach mouthA can now also reach mouthB's future light cone,
 *    and vice versa. The traversal cost is `traversalTimePLo`.
 * 4. For each VENT (spacelike) edge, check whether the wormhole makes
 *    the pair causally connected. If so, rewire it to PROCESS.
 * 5. Build the modified topology and compute connectivity change.
 */
export function insertWormhole(
  topology: CausalTopologyPacket,
  connection: WormholeConnection
): WormholeTopologyResult {
  const { mouthA, mouthB, traversalTimePLo } = connection;

  // Validate that both mouths exist in the topology
  const eventA = topology.events.find((e: SpacetimeEvent) => e.id === mouthA);
  const eventB = topology.events.find((e: SpacetimeEvent) => e.id === mouthB);
  if (!eventA) {
    throw new Error(`Wormhole mouth A event not found: ${mouthA}`);
  }
  if (!eventB) {
    throw new Error(`Wormhole mouth B event not found: ${mouthB}`);
  }

  // Build reachability maps from original causal edges
  const forward = buildForwardReach(topology.edges);
  const reverse = buildReverseReach(topology.edges);

  // Transitive reachability sets
  // pastOfA: all events that can causally reach mouthA (A is in their future)
  const pastOfA = transitiveReverseReach(mouthA, reverse);
  pastOfA.add(mouthA); // mouthA can reach itself trivially
  // futureOfA: all events mouthA can causally reach
  const futureOfA = transitiveForwardReach(mouthA, forward);
  futureOfA.add(mouthA);
  // pastOfB: all events that can causally reach mouthB
  const pastOfB = transitiveReverseReach(mouthB, reverse);
  pastOfB.add(mouthB);
  // futureOfB: all events mouthB can causally reach
  const futureOfB = transitiveForwardReach(mouthB, forward);
  futureOfB.add(mouthB);

  // The wormhole creates new causal paths:
  // Path A→B: X ∈ pastOfA ∪ {A} → wormhole → B → Y ∈ futureOfB ∪ {B}
  //   So (X, Y) becomes causally connected if X ∈ pastOfA and Y ∈ futureOfB
  // Path B→A: X ∈ pastOfB ∪ {B} → wormhole → A → Y ∈ futureOfA ∪ {A}
  //   So (X, Y) becomes causally connected if X ∈ pastOfB and Y ∈ futureOfA

  // Build the set of newly causal pairs by checking spacelike edges
  const rewiredEdges: RewiredEdge[] = [];
  const newlyCausalPairs: { readonly eventA: string; readonly eventB: string }[] = [];

  // Check which spacelike edges become causal through the wormhole
  const rewiredPairKeys = new Set<string>();

  for (const edge of topology.spacelikeEdges) {
    const src = edge.sourceId;
    const tgt = edge.targetId;

    // Check A→B path: src reaches A, B reaches tgt
    const viaAB =
      (pastOfA.has(src) && futureOfB.has(tgt)) ||
      (pastOfA.has(tgt) && futureOfB.has(src));

    // Check B→A path: src reaches B, A reaches tgt
    const viaBA =
      (pastOfB.has(src) && futureOfA.has(tgt)) ||
      (pastOfB.has(tgt) && futureOfA.has(src));

    // Direct wormhole connection: the two mouths themselves
    const directMouths =
      (src === mouthA && tgt === mouthB) ||
      (src === mouthB && tgt === mouthA);

    if (viaAB || viaBA || directMouths) {
      const pairKey = [src, tgt].sort().join('::');
      if (!rewiredPairKeys.has(pairKey)) {
        rewiredPairKeys.add(pairKey);
        rewiredEdges.push({
          sourceId: src,
          targetId: tgt,
          originalEdgeType: 'VENT',
          newEdgeType: 'PROCESS',
        });
        newlyCausalPairs.push({ eventA: src, eventB: tgt });
      }
    }
  }

  // Build modified edges: replace rewired VENT edges with PROCESS
  const modifiedEvents = [...topology.events];
  const modifiedEdges: CausalEdge[] = topology.edges.map((edge: CausalEdge) => {
    const pairKey = [edge.sourceId, edge.targetId].sort().join('::');
    if (edge.edgeType === 'VENT' && rewiredPairKeys.has(pairKey)) {
      return {
        ...edge,
        edgeType: 'PROCESS' as CausalEdgeType,
      };
    }
    return edge;
  });

  // Rebuild topology packet from modified edges
  const timelikeEdges: CausalEdge[] = [];
  const lightlikeEdges: CausalEdge[] = [];
  const spacelikeEdges: CausalEdge[] = [];

  for (const edge of modifiedEdges) {
    if (edge.edgeType === 'PROCESS') timelikeEdges.push(edge);
    else if (edge.edgeType === 'RACE') lightlikeEdges.push(edge);
    else spacelikeEdges.push(edge);
  }

  const totalPairs = modifiedEdges.length;
  const causalPairs = timelikeEdges.length + lightlikeEdges.length;
  const modifiedConnectivity = totalPairs > 0 ? causalPairs / totalPairs : 0;

  // Build summary for modified topology
  const spacelikeCountMap = new Map<string, number>();
  const causalCountMap = new Map<string, number>();
  let maxLightCrossing = 0;

  for (const e of modifiedEvents) {
    spacelikeCountMap.set(e.id, 0);
    causalCountMap.set(e.id, 0);
  }

  for (const edge of modifiedEdges) {
    if (edge.edgeType === 'VENT') {
      spacelikeCountMap.set(
        edge.sourceId,
        (spacelikeCountMap.get(edge.sourceId) ?? 0) + 1
      );
      spacelikeCountMap.set(
        edge.targetId,
        (spacelikeCountMap.get(edge.targetId) ?? 0) + 1
      );
    } else {
      causalCountMap.set(
        edge.sourceId,
        (causalCountMap.get(edge.sourceId) ?? 0) + 1
      );
      causalCountMap.set(
        edge.targetId,
        (causalCountMap.get(edge.targetId) ?? 0) + 1
      );
    }
    if (edge.interval.lightCrossingPLo > maxLightCrossing) {
      maxLightCrossing = edge.interval.lightCrossingPLo;
    }
  }

  let mostIsolatedId = modifiedEvents[0]?.id ?? '';
  let maxSpacelike = 0;
  let mostConnectedId = modifiedEvents[0]?.id ?? '';
  let maxCausal = 0;

  for (const e of modifiedEvents) {
    const sc = spacelikeCountMap.get(e.id) ?? 0;
    const cc = causalCountMap.get(e.id) ?? 0;
    if (sc > maxSpacelike) {
      maxSpacelike = sc;
      mostIsolatedId = e.id;
    }
    if (cc > maxCausal) {
      maxCausal = cc;
      mostConnectedId = e.id;
    }
  }

  const modifiedTopology: CausalTopologyPacket = {
    source: 'GnosisCausalTopology',
    events: modifiedEvents,
    edges: modifiedEdges,
    timelikeEdges,
    lightlikeEdges,
    spacelikeEdges,
    summary: {
      eventCount: modifiedEvents.length,
      edgeCount: modifiedEdges.length,
      timelikeCount: timelikeEdges.length,
      lightlikeCount: lightlikeEdges.length,
      spacelikeCount: spacelikeEdges.length,
      causalConnectivity: modifiedConnectivity,
      diameterPLo: maxLightCrossing,
      mostIsolatedEventId: mostIsolatedId,
      mostConnectedEventId: mostConnectedId,
    },
  };

  const originalConnectivity = topology.summary.causalConnectivity;

  return {
    source: 'GnosisWormholeTopology',
    connection,
    originalTopology: topology,
    modifiedTopology,
    rewiredEdges,
    newlyCausalPairs,
    originalCausalConnectivity: originalConnectivity,
    modifiedCausalConnectivity: modifiedConnectivity,
    connectivityGain: modifiedConnectivity - originalConnectivity,
  };
}

/**
 * Find all events now reachable from `eventId` through the wormhole
 * that were not reachable before.
 *
 * Starting from `eventId`, collect its transitive future causal reach.
 * If any of those events (or eventId itself) sit at a wormhole mouth,
 * add the other mouth's future causal reach. Return only the newly
 * added event IDs (those not in the original forward reach).
 */
export function wormholeTransitiveReach(
  topology: CausalTopologyPacket,
  connection: WormholeConnection,
  eventId: string
): readonly string[] {
  const forward = buildForwardReach(topology.edges);

  // Original transitive reach (without wormhole)
  const originalReach = transitiveForwardReach(eventId, forward);

  // Augmented reach: include wormhole shortcuts
  const augmentedReach = new Set(originalReach);
  const toExplore: string[] = [];

  // Check if eventId itself is a wormhole mouth
  if (eventId === connection.mouthA) {
    toExplore.push(connection.mouthB);
  } else if (eventId === connection.mouthB) {
    toExplore.push(connection.mouthA);
  }

  // Check if any originally reachable event is a wormhole mouth
  for (const reachable of originalReach) {
    if (reachable === connection.mouthA) {
      toExplore.push(connection.mouthB);
    } else if (reachable === connection.mouthB) {
      toExplore.push(connection.mouthA);
    }
  }

  // Expand from the other side of the wormhole
  for (const start of toExplore) {
    if (!augmentedReach.has(start)) {
      augmentedReach.add(start);
    }
    const otherSideReach = transitiveForwardReach(start, forward);
    for (const r of otherSideReach) {
      augmentedReach.add(r);
    }
  }

  // Also check the reverse direction: events reachable through
  // wormhole mouths that are in the augmented set
  // (second pass for mouths discovered in the expansion)
  for (const r of augmentedReach) {
    if (r === connection.mouthA && !augmentedReach.has(connection.mouthB)) {
      augmentedReach.add(connection.mouthB);
      const extra = transitiveForwardReach(connection.mouthB, forward);
      for (const e of extra) augmentedReach.add(e);
    }
    if (r === connection.mouthB && !augmentedReach.has(connection.mouthA)) {
      augmentedReach.add(connection.mouthA);
      const extra = transitiveForwardReach(connection.mouthA, forward);
      for (const e of extra) augmentedReach.add(e);
    }
  }

  // Return only the newly reachable events (exclude the starting event itself)
  const newlyReachable: string[] = [];
  for (const r of augmentedReach) {
    if (r !== eventId && !originalReach.has(r)) {
      newlyReachable.push(r);
    }
  }

  return newlyReachable;
}
