// @a0n/.ignored_gnosis/src/x-ranger-boundary.ts
import { sha256Hex, stableStringify } from './universe/index';

export const XRANGER_BOUNDARY_THEOREM_IDS = [
  'THM-MESH-MASKED-SOFTMAX-LINF-NONEXPANSIVE',
  'THM-MESH-DOBRUSHIN-OSC-CONTRACTION',
  'THM-BULEYEAN-DIST-UNIFORM-FLOOR',
  'THM-BULEYEAN-KERNEL-DOBRUSHIN-FLOOR',
  'THM-MESH-DISTRIBUTED-CERT-GLOBALIZES',
  'THM-MESH-OBSERVABILITY-COMPLETION-MONOTONE',
] as const;

export type XRangerBoundaryTheoremId =
  (typeof XRANGER_BOUNDARY_THEOREM_IDS)[number];

export const XRANGER_BOUNDARY_ACTIONS = [
  'allow',
  'mirror',
  'rate_limit',
  'quarantine',
  'snapshot',
  'drop',
] as const;

export type XRangerBoundaryAction =
  (typeof XRANGER_BOUNDARY_ACTIONS)[number];

export const XRANGER_INSPECTION_ACTIONS = [
  'mirror',
  'quarantine',
  'snapshot',
] as const;

export type XRangerInspectionAction =
  (typeof XRANGER_INSPECTION_ACTIONS)[number];

export type XRangerTransport =
  | 'tcp'
  | 'udp'
  | 'http'
  | 'https'
  | 'quic'
  | 'unix';

export type XRangerDirection = 'ingress' | 'egress' | 'lateral';

export type XRangerTrustTier =
  | 'public'
  | 'partner'
  | 'internal'
  | 'privileged'
  | 'control-plane';

export interface XRangerStateClass {
  readonly id: string;
  readonly localPortClass: string;
  readonly remotePortClass: string;
  readonly protocol: XRangerTransport;
  readonly direction: XRangerDirection;
  readonly trustTier: XRangerTrustTier;
}

export interface XRangerInspectionFloor {
  readonly action: XRangerInspectionAction;
  readonly numerator: number;
  readonly denominator: number;
}

export interface XRangerPrismShroud {
  readonly mode: 'none' | 'attested-route' | 'delayed-attested-route';
  readonly privacyModel: 'accountability';
}

export interface XRangerMeshPolicy {
  readonly stateClasses: readonly XRangerStateClass[];
  readonly actions: readonly XRangerBoundaryAction[];
  readonly inspectionFloors: readonly XRangerInspectionFloor[];
  readonly certificationMode: 'local-quorum-globalized';
  readonly reconstructionMode: 'spec-boundary';
  readonly prismShroud: XRangerPrismShroud;
}

export interface XRangerBoundaryShellContract {
  readonly packageRoot: string;
  readonly role: 'package-shell';
  readonly compatibility: 'wrangler-compatible';
  readonly emitsDeploymentManifest: boolean;
}

export interface XRangerBoundaryManagerContract {
  readonly packageRoot: string;
  readonly role: 'rust-manager';
  readonly transportFront: 'gnosis-uring';
  readonly workerRuntime: string;
  readonly emitsForgeDeploymentManifest: boolean;
  readonly futureTarget: 'lambda';
}

export interface XRangerBoundaryManifest {
  readonly kind: 'x-ranger-boundary';
  readonly version: 1;
  readonly generatedAt: string;
  readonly digest: string;
  readonly theoremIds: readonly XRangerBoundaryTheoremId[];
  readonly shell: XRangerBoundaryShellContract;
  readonly manager: XRangerBoundaryManagerContract;
  readonly policy: XRangerMeshPolicy;
}

export interface CreateXRangerBoundaryManifestInput {
  readonly generatedAt?: string;
  readonly theoremIds?: readonly XRangerBoundaryTheoremId[];
  readonly shell?: Partial<XRangerBoundaryShellContract>;
  readonly manager?: Partial<XRangerBoundaryManagerContract>;
  readonly policy: XRangerMeshPolicy;
}

export interface XRangerBoundaryManifestAudit {
  readonly selfConsistent: boolean;
  readonly issues: readonly string[];
  readonly weakestInspectionFloor: number | null;
}

const ACTION_ORDER: readonly XRangerBoundaryAction[] = XRANGER_BOUNDARY_ACTIONS;
const INSPECTION_ACTION_SET = new Set<string>(XRANGER_INSPECTION_ACTIONS);

function nowIso(): string {
  return new Date().toISOString();
}

function trimOrFallback(value: string | undefined, fallback: string): string {
  const normalized = value?.trim();
  return normalized && normalized.length > 0 ? normalized : fallback;
}

function clampPositiveInt(value: number): number {
  if (!Number.isFinite(value)) {
    return 1;
  }
  return Math.max(1, Math.floor(value));
}

function actionRank(action: XRangerBoundaryAction): number {
  return ACTION_ORDER.indexOf(action);
}

function inspectionFloorProbability(floor: XRangerInspectionFloor): number {
  return floor.numerator / floor.denominator;
}

function sortStateClasses(
  stateClasses: readonly XRangerStateClass[]
): XRangerStateClass[] {
  return [...stateClasses].sort((left, right) => left.id.localeCompare(right.id));
}

function normalizeStateClass(stateClass: XRangerStateClass): XRangerStateClass {
  return {
    id: trimOrFallback(stateClass.id, 'state'),
    localPortClass: trimOrFallback(stateClass.localPortClass, 'any'),
    remotePortClass: trimOrFallback(stateClass.remotePortClass, 'any'),
    protocol: stateClass.protocol,
    direction: stateClass.direction,
    trustTier: stateClass.trustTier,
  };
}

function dedupeActions(
  actions: readonly XRangerBoundaryAction[]
): XRangerBoundaryAction[] {
  const seen = new Set<XRangerBoundaryAction>();
  const normalized: XRangerBoundaryAction[] = [];
  for (const action of actions) {
    if (seen.has(action)) {
      continue;
    }
    seen.add(action);
    normalized.push(action);
  }
  return normalized.sort((left, right) => actionRank(left) - actionRank(right));
}

function strongestFloor(
  left: XRangerInspectionFloor,
  right: XRangerInspectionFloor
): XRangerInspectionFloor {
  return left.numerator * right.denominator >= right.numerator * left.denominator
    ? left
    : right;
}

function normalizeInspectionFloors(
  floors: readonly XRangerInspectionFloor[]
): XRangerInspectionFloor[] {
  const strongestByAction = new Map<
    XRangerInspectionAction,
    XRangerInspectionFloor
  >();

  for (const floor of floors) {
    const normalized: XRangerInspectionFloor = {
      action: floor.action,
      numerator: clampPositiveInt(floor.numerator),
      denominator: clampPositiveInt(floor.denominator),
    };
    const current = strongestByAction.get(normalized.action);
    strongestByAction.set(
      normalized.action,
      current ? strongestFloor(current, normalized) : normalized
    );
  }

  return [...strongestByAction.values()].sort(
    (left, right) => actionRank(left.action) - actionRank(right.action)
  );
}

function normalizeTheoremIds(
  theoremIds: readonly XRangerBoundaryTheoremId[] | undefined
): readonly XRangerBoundaryTheoremId[] {
  const ids =
    theoremIds && theoremIds.length > 0
      ? [...new Set(theoremIds)]
      : [...XRANGER_BOUNDARY_THEOREM_IDS];
  return ids.sort((left, right) => left.localeCompare(right));
}

function normalizePolicy(policy: XRangerMeshPolicy): XRangerMeshPolicy {
  return {
    stateClasses: sortStateClasses(policy.stateClasses.map(normalizeStateClass)),
    actions: dedupeActions(policy.actions),
    inspectionFloors: normalizeInspectionFloors(policy.inspectionFloors),
    certificationMode: policy.certificationMode,
    reconstructionMode: policy.reconstructionMode,
    prismShroud: policy.prismShroud,
  };
}

function digestManifestBody(
  body: Omit<XRangerBoundaryManifest, 'digest'>
): string {
  return sha256Hex(stableStringify(body));
}

/**
 * Handles the gnosis floor Probability workflow.
 */
export function floorProbability(floor: XRangerInspectionFloor): number {
  return inspectionFloorProbability(floor);
}

/**
 * Handles the gnosis weakest Inspection Floor workflow.
 */
export function weakestInspectionFloor(
  floors: readonly XRangerInspectionFloor[]
): number | null {
  if (floors.length === 0) {
    return null;
  }
  return Math.min(...floors.map(inspectionFloorProbability));
}

/**
 * Creates the XRanger Boundary Manifest.
 */
export function createXRangerBoundaryManifest(
  input: CreateXRangerBoundaryManifestInput
): XRangerBoundaryManifest {
  const generatedAt = trimOrFallback(input.generatedAt, nowIso());
  const theoremIds = normalizeTheoremIds(input.theoremIds);
  const shell: XRangerBoundaryShellContract = {
    packageRoot: trimOrFallback(
      input.shell?.packageRoot,
      'open-source/x-ranger'
    ),
    role: 'package-shell',
    compatibility: 'wrangler-compatible',
    emitsDeploymentManifest: input.shell?.emitsDeploymentManifest ?? true,
  };
  const manager: XRangerBoundaryManagerContract = {
    packageRoot: trimOrFallback(
      input.manager?.packageRoot,
      'open-source/x-ranger/gnosis-ranger'
    ),
    role: 'rust-manager',
    transportFront: 'gnosis-uring',
    workerRuntime: trimOrFallback(
      input.manager?.workerRuntime,
      'local-gnosis-worker'
    ),
    emitsForgeDeploymentManifest:
      input.manager?.emitsForgeDeploymentManifest ?? true,
    futureTarget: 'lambda',
  };
  const policy = normalizePolicy(input.policy);

  const body = {
    kind: 'x-ranger-boundary' as const,
    version: 1 as const,
    generatedAt,
    theoremIds,
    shell,
    manager,
    policy,
  };

  return {
    ...body,
    digest: digestManifestBody(body),
  };
}

/**
 * Handles the gnosis audit XRanger Boundary Manifest workflow.
 */
export function auditXRangerBoundaryManifest(
  manifest: XRangerBoundaryManifest
): XRangerBoundaryManifestAudit {
  const issues: string[] = [];

  const expectedDigest = digestManifestBody({
    kind: manifest.kind,
    version: manifest.version,
    generatedAt: manifest.generatedAt,
    theoremIds: manifest.theoremIds,
    shell: manifest.shell,
    manager: manifest.manager,
    policy: manifest.policy,
  });
  if (manifest.digest !== expectedDigest) {
    issues.push('digest does not match the canonical x-ranger boundary body');
  }

  if (manifest.shell.packageRoot.trim().length === 0) {
    issues.push('shell packageRoot must be non-empty');
  }
  if (!manifest.shell.emitsDeploymentManifest) {
    issues.push('shell must emit a deployment manifest at the x-ranger boundary');
  }

  if (manifest.manager.packageRoot.trim().length === 0) {
    issues.push('manager packageRoot must be non-empty');
  }
  if (manifest.manager.packageRoot === manifest.shell.packageRoot) {
    issues.push('shell and manager package roots must remain distinct');
  }
  if (manifest.manager.workerRuntime.trim().length === 0) {
    issues.push('manager workerRuntime must be non-empty');
  }
  if (!manifest.manager.emitsForgeDeploymentManifest) {
    issues.push(
      'manager must emit the Forge-friendly deployment manifest for the future lambda path'
    );
  }

  if (manifest.policy.stateClasses.length === 0) {
    issues.push('policy must declare at least one state class');
  }

  const seenStateIds = new Set<string>();
  for (const stateClass of manifest.policy.stateClasses) {
    if (stateClass.id.trim().length === 0) {
      issues.push('policy state class ids must be non-empty');
      continue;
    }
    if (seenStateIds.has(stateClass.id)) {
      issues.push(`duplicate policy state class id '${stateClass.id}'`);
      continue;
    }
    seenStateIds.add(stateClass.id);
  }

  if (manifest.policy.actions.length === 0) {
    issues.push('policy must expose at least one action');
  }

  if (manifest.policy.inspectionFloors.length === 0) {
    issues.push('policy must declare at least one inspection floor');
  }

  const actionSet = new Set(manifest.policy.actions);
  for (const floor of manifest.policy.inspectionFloors) {
    if (!INSPECTION_ACTION_SET.has(floor.action)) {
      issues.push(`inspection floor action '${floor.action}' is not inspectable`);
    }
    if (!actionSet.has(floor.action)) {
      issues.push(
        `inspection floor action '${floor.action}' must also appear in policy actions`
      );
    }
    if (floor.numerator < 1) {
      issues.push(`inspection floor '${floor.action}' must have numerator >= 1`);
    }
    if (floor.denominator < 1) {
      issues.push(
        `inspection floor '${floor.action}' must have denominator >= 1`
      );
    }
    if (floor.numerator > floor.denominator) {
      issues.push(
        `inspection floor '${floor.action}' cannot exceed probability 1`
      );
    }
  }

  const weakestFloor = weakestInspectionFloor(manifest.policy.inspectionFloors);

  return {
    selfConsistent: issues.length === 0,
    issues,
    weakestInspectionFloor: weakestFloor,
  };
}
