// @a0n/gnosis/distributed-inference-host/src/test-fixtures/local-mesh.ts
//! Local mesh fixture for the consumption-wave bench.
//!
//! Constructs a Pipeline pointed at two local fat-station instances
//! (typically launched by scripts/spawn-local-mesh.sh on ports 8081
//! and 8082). The fixture exists for offline / smoke measurement of
//! the Amplituhedron coordinator + bw-codec wire WITHOUT a deployed
//! Cloudflare Worker mesh.
//!
//! Note on shape: the actual `Station` config (see ../station.ts) uses
//! `startLayer`/`endLayer` — the original sketch's `layerLo`/`layerHi`
//! was wrong and is adapted here. Pipeline (see ../pipeline.ts) takes
//! `hiddenDim` (required) and has no `model` field; the model name is
//! captured in `LocalMeshConfig` purely as bench-harness metadata so
//! callers can stamp telemetry, NOT because Pipeline reads it.

import { Station } from '../station';
import { Pipeline } from '../pipeline';

export interface LocalMeshConfig {
  /** Defaults to `http://localhost:8081`. */
  stationAUrl?: string;
  /** Defaults to `http://localhost:8082`. */
  stationBUrl?: string;
  /**
   * Bench-harness metadata only — Pipeline itself does not consume a
   * model name. Defaults to `qwen-coder-7b`. Returned alongside the
   * Pipeline so the caller can stamp the metric/log lines.
   */
  model?: string;
  /** Layer count for the model. Defaults to 28 (qwen-coder-7b). */
  totalLayers?: number;
  /**
   * Hidden dimension for the model. Required by Pipeline so it can
   * slice contiguous residual chunks. Defaults to 4096 (qwen-coder-7b).
   */
  hiddenDim?: number;
  /**
   * Per-station batch capacity hint. The local fat-station native
   * server runs `BATCH_CAPACITY = 8`; the wasm worker caps at 4.
   * Defaults to 8 because this fixture is documented as pointing at
   * native fat-station instances.
   */
  maxBatchSize?: number;
}

export interface LocalMeshBundle {
  pipeline: Pipeline;
  stationA: Station;
  stationB: Station;
  /** Echoed back from config so the caller can stamp telemetry. */
  model: string;
  stationAUrl: string;
  stationBUrl: string;
  totalLayers: number;
  midLayer: number;
}

/**
 * Build a Pipeline + the two underlying Stations pointed at local
 * fat-station instances. The Pipeline is the load-bearing return; the
 * stations are surfaced so a bench harness can probe per-station
 * state (`startLayer`/`endLayer`/`role`) without re-deriving it.
 */
export function createLocalMeshPipeline(config: LocalMeshConfig = {}): LocalMeshBundle {
  const stationAUrl = config.stationAUrl ?? 'http://localhost:8081';
  const stationBUrl = config.stationBUrl ?? 'http://localhost:8082';
  const totalLayers = config.totalLayers ?? 28;
  const hiddenDim = config.hiddenDim ?? 4096;
  const maxBatchSize = config.maxBatchSize ?? 8;
  const model = config.model ?? 'qwen-coder-7b';

  if (!Number.isInteger(totalLayers) || totalLayers < 2) {
    throw new Error(
      `createLocalMeshPipeline: totalLayers must be an integer >= 2 (got ${String(totalLayers)})`,
    );
  }

  const midLayer = Math.floor(totalLayers / 2);

  const stationA = new Station({
    endpoint: stationAUrl,
    startLayer: 0,
    endLayer: midLayer,
    role: 'entry',
    maxBatchSize,
  });

  const stationB = new Station({
    endpoint: stationBUrl,
    startLayer: midLayer,
    endLayer: totalLayers,
    role: 'exit',
    maxBatchSize,
  });

  const pipeline = new Pipeline({
    stations: [stationA, stationB],
    hiddenDim,
  });

  return {
    pipeline,
    stationA,
    stationB,
    model,
    stationAUrl,
    stationBUrl,
    totalLayers,
    midLayer,
  };
}

/**
 * Probe both stations for liveness. Returns true iff both respond OK.
 * Never throws — a network error / DNS miss / connection refused all
 * resolve to `false`. Useful as a precondition gate for bench runs.
 */
export async function pingLocalMesh(config: LocalMeshConfig = {}): Promise<boolean> {
  const a = config.stationAUrl ?? 'http://localhost:8081';
  const b = config.stationBUrl ?? 'http://localhost:8082';
  try {
    const [resA, resB] = await Promise.all([
      fetch(`${a}/health`, { method: 'GET' }),
      fetch(`${b}/health`, { method: 'GET' }),
    ]);
    return resA.ok && resB.ok;
  } catch {
    return false;
  }
}
