// @a0n/gnosis/frffmpeg/js/frffmpeg-wasm.ts
/**
 * frffmpeg-wasm — ergonomic, runtime-agnostic bridge to the sovereign frffmpeg WASM encoder.
 *
 * The bundled artifact `frffmpeg.wasm` is regenerated from the Rust crate with:
 *   cargo build --release --target wasm32-unknown-unknown -p frffmpeg --lib
 * then copied here as the stable, committable path. The Worker/test always import this
 * fixed artifact, matching the pattern used by the social-organ WASM vendor.
 *
 * Runtime requirements: any environment with a Web-standard `WebAssembly` global (Node ≥16,
 * Cloudflare Workers, Deno, browsers). No node:fs usage in this module — the caller
 * supplies raw wasm bytes so this file is fully Worker-safe.
 *
 * Public API
 * ----------
 * loadFrffmpeg(wasmSource)         → Promise<Frffmpeg>
 * frffmpeg.encodeRgbFramesToMp4(frames, width, height, fps) → Uint8Array
 */

// ---------------------------------------------------------------------------
// WASM exports shape
// ---------------------------------------------------------------------------

interface FrffmpegExports {
  /** Linear memory shared between host and wasm. */
  memory: WebAssembly.Memory;
  /** Allocate `len` bytes and return a pointer. Free with frffmpeg_free. */
  frffmpeg_alloc(len: number): number;
  /** Free a buffer previously returned by alloc or encode. */
  frffmpeg_free(ptr: number, len: number): void;
  /**
   * Encode `n_frames` contiguous RGB24 frames to an all-intra faststart MP4.
   * Writes the output length to `*out_len_ptr` and returns a pointer to the MP4 bytes.
   * Returns 0 on bad input.
   */
  frffmpeg_encode_ipcm(
    frames_ptr: number,
    frame_len: number,
    n_frames: number,
    width: number,
    height: number,
    fps: number,
    out_len_ptr: number,
  ): number;
  /**
   * Encode `n_frames` contiguous RGB24 frames to an all-intra COMPRESSED faststart MP4
   * (real intra-prediction + CAVLC, 12–78× smaller than I_PCM). `qp` is the quantiser.
   * Writes the output length to `*out_len_ptr` and returns a pointer to the MP4 bytes.
   * Returns 0 on bad input.
   */
  frffmpeg_encode_compressed(
    frames_ptr: number,
    frame_len: number,
    n_frames: number,
    width: number,
    height: number,
    fps: number,
    qp: number,
    out_len_ptr: number,
  ): number;
  /**
   * Encode `n_frames` contiguous RGB24 frames to an all-intra COMPRESSED faststart MP4
   * whose per-frame QP ADAPTS to hit `target_kbps` (target video bitrate). Additive
   * companion to `frffmpeg_encode_compressed` (fixed qp). Writes the output length to
   * `*out_len_ptr` and returns a pointer to the MP4 bytes. Returns 0 on bad input.
   */
  frffmpeg_encode_rate_controlled(
    frames_ptr: number,
    frame_len: number,
    n_frames: number,
    width: number,
    height: number,
    fps: number,
    target_kbps: number,
    out_len_ptr: number,
  ): number;
  /**
   * Rate-controlled COMPRESSED video (`target_kbps`) + interleaved i16 PCM (AAC-LC audio)
   * into ONE faststart MP4 with two tracks. `pcm_ptr` points at `pcm_samples` i16 values.
   * Writes the MP4 length to `*out_len_ptr` and returns a pointer to the MP4 bytes.
   * Returns 0 on bad input.
   */
  frffmpeg_encode_rate_controlled_av(
    frames_ptr: number,
    frame_len: number,
    n_frames: number,
    width: number,
    height: number,
    fps: number,
    target_kbps: number,
    pcm_ptr: number,
    pcm_samples: number,
    sample_rate: number,
    channels: number,
    out_len_ptr: number,
  ): number;
  /**
   * Encode ONE RGB24 frame into a single COMPRESSED AVCC sample (the fanout/shard unit).
   * `idr_pic_id` distinguishes the IDR. Writes the sample length to `*out_len_ptr` and
   * returns a pointer to the sample bytes. Returns 0 on bad input.
   */
  frffmpeg_encode_sample_compressed(
    frame_ptr: number,
    frame_len: number,
    width: number,
    height: number,
    qp: number,
    idr_pic_id: number,
    out_len_ptr: number,
  ): number;
  /**
   * Encode `n_frames` RGB24 frames into rate-controlled COMPRESSED AVCC samples (one IDR per
   * frame, per-frame QP adapting to `target_kbps`) in ONE call. `frames_ptr` is all frames
   * concatenated; `lens_ptr` is a host-allocated array of `n_frames` writable u32s that the
   * call fills with each frame's sample length (frame order). Writes the concatenated-blob
   * length to `*out_len_ptr` and returns a pointer to the blob. Returns 0 on bad input.
   */
  frffmpeg_encode_rate_controlled_samples(
    frames_ptr: number,
    frame_len: number,
    n_frames: number,
    width: number,
    height: number,
    fps: number,
    target_kbps: number,
    lens_ptr: number,
    out_len_ptr: number,
  ): number;
  /**
   * Encode `n_frames` RGB24 frames (COMPRESSED video) + interleaved i16 PCM (AAC-LC audio)
   * into ONE faststart MP4 with two tracks. `pcm_ptr` points at `pcm_samples` i16 values.
   * Writes the MP4 length to `*out_len_ptr` and returns a pointer to the MP4 bytes.
   * Returns 0 on bad input.
   */
  frffmpeg_encode_av(
    frames_ptr: number,
    frame_len: number,
    n_frames: number,
    width: number,
    height: number,
    fps: number,
    qp: number,
    pcm_ptr: number,
    pcm_samples: number,
    sample_rate: number,
    channels: number,
    out_len_ptr: number,
  ): number;
  /**
   * Mux pre-encoded shard samples (the fold step) into a faststart MP4. `samples_ptr` is
   * all sample bytes concatenated; `lens_ptr` is an array of `n_samples` u32 lengths.
   * Writes the MP4 length to `*out_len_ptr` and returns a pointer to the MP4 bytes.
   * Returns 0 on bad input.
   */
  frffmpeg_mux_samples(
    samples_ptr: number,
    lens_ptr: number,
    n_samples: number,
    width: number,
    height: number,
    fps: number,
    out_len_ptr: number,
  ): number;
  /**
   * Mux pre-encoded shard video samples (the fold step) TOGETHER with an AAC-LC audio
   * track into a faststart MP4 with TWO tracks — so SHARDED (fanout) clips carry sound.
   * `samples_ptr` is all sample bytes concatenated; `lens_ptr` is `n_samples` u32 lengths;
   * `pcm_ptr` points at `pcm_samples` interleaved i16 values. Writes the MP4 length to
   * `*out_len_ptr` and returns a pointer to the MP4 bytes. Returns 0 on bad input.
   */
  frffmpeg_mux_samples_av(
    samples_ptr: number,
    lens_ptr: number,
    n_samples: number,
    width: number,
    height: number,
    fps: number,
    pcm_ptr: number,
    pcm_samples: number,
    sample_rate: number,
    channels: number,
    out_len_ptr: number,
  ): number;
}

// ---------------------------------------------------------------------------
// Frffmpeg class
// ---------------------------------------------------------------------------

/**
 * An instantiated frffmpeg WASM encoder. Obtain via `loadFrffmpeg`.
 */
export class Frffmpeg {
  /** @internal */
  readonly exports: FrffmpegExports;

  /** @internal */
  constructor(exports: FrffmpegExports) {
    this.exports = exports;
  }

  /**
   * Encode an array of packed RGB24 frames into an all-intra faststart MP4.
   *
   * @param frames - Each element is exactly `width * height * 3` bytes of packed RGB24.
   * @param width  - Frame width in pixels.
   * @param height - Frame height in pixels.
   * @param fps    - Playback frame rate.
   * @returns A COPIED Uint8Array containing the complete MP4 file.
   *
   * Memory-grow safety: the encode call internally allocates and CAN grow wasm linear
   * memory, which DETACHES any previously-cached `memory.buffer` ArrayBuffer. We
   * re-acquire `new Uint8Array(this.exports.memory.buffer)` after every alloc and
   * after the encode call, immediately before each read/write.
   */
  encodeRgbFramesToMp4(
    frames: Uint8Array[],
    width: number,
    height: number,
    fps: number,
  ): Uint8Array {
    const { exports } = this;
    const n = frames.length;
    if (n === 0) throw new RangeError("frames must be non-empty");

    const frameLen = width * height * 3;
    for (let i = 0; i < n; i++) {
      if (frames[i].length !== frameLen) {
        throw new RangeError(
          `frame[${i}] has ${frames[i].length} bytes; expected ${frameLen} (${width}×${height}×3)`,
        );
      }
    }

    const totalLen = frameLen * n;

    // --- allocate input buffer ---
    const inPtr = exports.frffmpeg_alloc(totalLen);
    // Re-acquire mem view: alloc may have grown memory.
    let mem = new Uint8Array(exports.memory.buffer);
    let offset = inPtr;
    for (let i = 0; i < n; i++) {
      mem.set(frames[i], offset);
      offset += frameLen;
    }

    // --- allocate 4-byte out_len cell (usize on wasm32 = u32 = 4 bytes) ---
    const outLenPtr = exports.frffmpeg_alloc(4);
    // Re-acquire after second alloc.
    mem = new Uint8Array(exports.memory.buffer);
    // Zero the out_len cell defensively.
    mem[outLenPtr] = 0;
    mem[outLenPtr + 1] = 0;
    mem[outLenPtr + 2] = 0;
    mem[outLenPtr + 3] = 0;

    // --- call encoder ---
    const outPtr = exports.frffmpeg_encode_ipcm(
      inPtr,
      frameLen,
      n,
      width,
      height,
      fps,
      outLenPtr,
    );

    // Re-acquire AFTER the encode call: it internally allocs and can grow memory.
    const outLen = new DataView(exports.memory.buffer).getUint32(outLenPtr, true);

    let result: Uint8Array;
    if (outPtr === 0 || outLen === 0) {
      result = new Uint8Array(0);
    } else {
      // Copy before freeing so the caller owns a stable buffer.
      result = new Uint8Array(
        exports.memory.buffer,
        outPtr,
        outLen,
      ).slice();
      exports.frffmpeg_free(outPtr, outLen);
    }

    // Free the input buffer and out_len cell.
    exports.frffmpeg_free(inPtr, totalLen);
    exports.frffmpeg_free(outLenPtr, 4);

    return result;
  }

  /**
   * Encode an array of packed RGB24 frames into an all-intra COMPRESSED faststart MP4
   * (real intra-prediction + CAVLC, 12–78× smaller than the I_PCM path).
   *
   * @param frames - Each element is exactly `width * height * 3` bytes of packed RGB24.
   * @param width  - Frame width in pixels.
   * @param height - Frame height in pixels.
   * @param fps    - Playback frame rate.
   * @param qp     - Quantiser parameter (0–51; lower = higher quality/bigger). Default 26.
   * @returns A COPIED Uint8Array containing the complete MP4 file.
   *
   * Memory-grow safety: identical to `encodeRgbFramesToMp4` — re-acquire
   * `new Uint8Array(this.exports.memory.buffer)` after every alloc and after the encode
   * call, immediately before each read/write.
   */
  encodeRgbFramesToMp4Compressed(
    frames: Uint8Array[],
    width: number,
    height: number,
    fps: number,
    qp = 26,
  ): Uint8Array {
    const { exports } = this;
    const n = frames.length;
    if (n === 0) throw new RangeError("frames must be non-empty");

    const frameLen = width * height * 3;
    for (let i = 0; i < n; i++) {
      if (frames[i].length !== frameLen) {
        throw new RangeError(
          `frame[${i}] has ${frames[i].length} bytes; expected ${frameLen} (${width}×${height}×3)`,
        );
      }
    }

    const totalLen = frameLen * n;

    // --- allocate input buffer ---
    const inPtr = exports.frffmpeg_alloc(totalLen);
    // Re-acquire mem view: alloc may have grown memory.
    let mem = new Uint8Array(exports.memory.buffer);
    let offset = inPtr;
    for (let i = 0; i < n; i++) {
      mem.set(frames[i], offset);
      offset += frameLen;
    }

    // --- allocate 4-byte out_len cell ---
    const outLenPtr = exports.frffmpeg_alloc(4);
    // Re-acquire after second alloc.
    mem = new Uint8Array(exports.memory.buffer);
    mem[outLenPtr] = 0;
    mem[outLenPtr + 1] = 0;
    mem[outLenPtr + 2] = 0;
    mem[outLenPtr + 3] = 0;

    // --- call encoder ---
    const outPtr = exports.frffmpeg_encode_compressed(
      inPtr,
      frameLen,
      n,
      width,
      height,
      fps,
      qp,
      outLenPtr,
    );

    // Re-acquire AFTER the encode call: it internally allocs and can grow memory.
    const outLen = new DataView(exports.memory.buffer).getUint32(outLenPtr, true);

    let result: Uint8Array;
    if (outPtr === 0 || outLen === 0) {
      result = new Uint8Array(0);
    } else {
      // Copy before freeing so the caller owns a stable buffer.
      result = new Uint8Array(
        exports.memory.buffer,
        outPtr,
        outLen,
      ).slice();
      exports.frffmpeg_free(outPtr, outLen);
    }

    // Free the input buffer and out_len cell.
    exports.frffmpeg_free(inPtr, totalLen);
    exports.frffmpeg_free(outLenPtr, 4);

    return result;
  }

  /**
   * Encode an array of packed RGB24 frames into a COMPRESSED faststart MP4 whose per-frame
   * QP ADAPTS to hit `targetKbps` (target video bitrate). One-pass per-frame feedback: every
   * frame is an independent IDR, so the QP chosen for each frame from the prior frame's
   * measured size never corrupts a later frame. QP is clamped to a sane range (18–40) so
   * quality never collapses and the stream never bloats to near-lossless.
   *
   * @param frames    - Each element is exactly `width * height * 3` bytes of packed RGB24.
   * @param width     - Frame width in pixels.
   * @param height    - Frame height in pixels.
   * @param fps       - Playback frame rate.
   * @param targetKbps - Target VIDEO bitrate in kbit/s (e.g. 800).
   * @returns A COPIED Uint8Array containing the complete MP4 file.
   */
  encodeRgbFramesToMp4RateControlled(
    frames: Uint8Array[],
    width: number,
    height: number,
    fps: number,
    targetKbps: number,
  ): Uint8Array {
    const { exports } = this;
    const n = frames.length;
    if (n === 0) throw new RangeError("frames must be non-empty");

    const frameLen = width * height * 3;
    for (let i = 0; i < n; i++) {
      if (frames[i].length !== frameLen) {
        throw new RangeError(
          `frame[${i}] has ${frames[i].length} bytes; expected ${frameLen} (${width}×${height}×3)`,
        );
      }
    }

    const totalLen = frameLen * n;

    // --- allocate input buffer ---
    const inPtr = exports.frffmpeg_alloc(totalLen);
    // Re-acquire mem view: alloc may have grown memory.
    let mem = new Uint8Array(exports.memory.buffer);
    let offset = inPtr;
    for (let i = 0; i < n; i++) {
      mem.set(frames[i], offset);
      offset += frameLen;
    }

    // --- allocate 4-byte out_len cell ---
    const outLenPtr = exports.frffmpeg_alloc(4);
    // Re-acquire after second alloc.
    mem = new Uint8Array(exports.memory.buffer);
    mem[outLenPtr] = 0;
    mem[outLenPtr + 1] = 0;
    mem[outLenPtr + 2] = 0;
    mem[outLenPtr + 3] = 0;

    // --- call rate-controlled encoder ---
    const outPtr = exports.frffmpeg_encode_rate_controlled(
      inPtr,
      frameLen,
      n,
      width,
      height,
      fps,
      targetKbps,
      outLenPtr,
    );

    // Re-acquire AFTER the encode call: it internally allocs and can grow memory.
    const outLen = new DataView(exports.memory.buffer).getUint32(outLenPtr, true);

    let result: Uint8Array;
    if (outPtr === 0 || outLen === 0) {
      result = new Uint8Array(0);
    } else {
      // Copy before freeing so the caller owns a stable buffer.
      result = new Uint8Array(exports.memory.buffer, outPtr, outLen).slice();
      exports.frffmpeg_free(outPtr, outLen);
    }

    // Free the input buffer and out_len cell.
    exports.frffmpeg_free(inPtr, totalLen);
    exports.frffmpeg_free(outLenPtr, 4);

    return result;
  }

  /**
   * Rate-controlled COMPRESSED video (`targetKbps`) together with interleaved i16 PCM
   * (AAC-LC audio) into ONE faststart MP4 with TWO tracks — a bitrate-targeted clip WITH
   * sound. Same per-frame QP feedback as `encodeRgbFramesToMp4RateControlled`.
   *
   * @param frames     - Each element is exactly `width * height * 3` bytes of packed RGB24.
   * @param width      - Frame width in pixels.
   * @param height     - Frame height in pixels.
   * @param fps        - Playback frame rate.
   * @param targetKbps - Target VIDEO bitrate in kbit/s (audio uses the AAC encoder's rate).
   * @param pcm        - Interleaved i16 PCM (mono: one value/sample; stereo: L,R,L,R…).
   * @param sampleRate - Audio sample rate in Hz (e.g. 44100).
   * @param channels   - 1 (mono) or 2 (stereo).
   * @returns A COPIED Uint8Array containing the complete two-track MP4 file.
   */
  encodeRgbFramesWithAudioToMp4RateControlled(
    frames: Uint8Array[],
    width: number,
    height: number,
    fps: number,
    targetKbps: number,
    pcm: Int16Array,
    sampleRate: number,
    channels: number,
  ): Uint8Array {
    const { exports } = this;
    const n = frames.length;
    if (n === 0) throw new RangeError("frames must be non-empty");

    const frameLen = width * height * 3;
    for (let i = 0; i < n; i++) {
      if (frames[i].length !== frameLen) {
        throw new RangeError(
          `frame[${i}] has ${frames[i].length} bytes; expected ${frameLen} (${width}×${height}×3)`,
        );
      }
    }

    const totalLen = frameLen * n;
    const pcmSamples = pcm.length;
    const pcmBytes = pcmSamples * 2; // i16 each

    // --- allocate input video buffer ---
    const inPtr = exports.frffmpeg_alloc(totalLen);
    let mem = new Uint8Array(exports.memory.buffer);
    let offset = inPtr;
    for (let i = 0; i < n; i++) {
      mem.set(frames[i], offset);
      offset += frameLen;
    }

    // --- allocate PCM buffer and write i16 little-endian ---
    const pcmPtr = exports.frffmpeg_alloc(pcmBytes > 0 ? pcmBytes : 1);
    let dv = new DataView(exports.memory.buffer);
    for (let i = 0; i < pcmSamples; i++) {
      dv.setInt16(pcmPtr + i * 2, pcm[i], true);
    }

    // --- allocate 4-byte out_len cell ---
    const outLenPtr = exports.frffmpeg_alloc(4);
    mem = new Uint8Array(exports.memory.buffer);
    mem[outLenPtr] = 0;
    mem[outLenPtr + 1] = 0;
    mem[outLenPtr + 2] = 0;
    mem[outLenPtr + 3] = 0;

    // --- call rate-controlled A/V encoder ---
    const outPtr = exports.frffmpeg_encode_rate_controlled_av(
      inPtr,
      frameLen,
      n,
      width,
      height,
      fps,
      targetKbps,
      pcmPtr,
      pcmSamples,
      sampleRate,
      channels,
      outLenPtr,
    );

    // Re-acquire AFTER the encode call: it internally allocs and can grow memory.
    const outLen = new DataView(exports.memory.buffer).getUint32(outLenPtr, true);

    let result: Uint8Array;
    if (outPtr === 0 || outLen === 0) {
      result = new Uint8Array(0);
    } else {
      result = new Uint8Array(exports.memory.buffer, outPtr, outLen).slice();
      exports.frffmpeg_free(outPtr, outLen);
    }

    exports.frffmpeg_free(inPtr, totalLen);
    exports.frffmpeg_free(pcmPtr, pcmBytes > 0 ? pcmBytes : 1);
    exports.frffmpeg_free(outLenPtr, 4);

    return result;
  }

  /**
   * Encode packed RGB24 frames (COMPRESSED video) together with interleaved i16 PCM
   * (AAC-LC audio) into ONE faststart MP4 with TWO tracks — a clip WITH sound.
   *
   * @param frames     - Each element is exactly `width * height * 3` bytes of packed RGB24.
   * @param width      - Frame width in pixels.
   * @param height     - Frame height in pixels.
   * @param fps        - Playback frame rate.
   * @param pcm        - Interleaved i16 PCM (mono: one value/sample; stereo: L,R,L,R…).
   * @param sampleRate - Audio sample rate in Hz (e.g. 44100).
   * @param channels   - 1 (mono) or 2 (stereo).
   * @param qp         - Video quantiser (0–51; lower = higher quality/bigger). Default 26.
   * @returns A COPIED Uint8Array containing the complete two-track MP4 file.
   *
   * Memory-grow safety: re-acquire `memory.buffer` views after every alloc and after the
   * encode call, immediately before each read/write. The PCM is written little-endian i16
   * directly into wasm linear memory.
   */
  encodeRgbFramesWithAudioToMp4(
    frames: Uint8Array[],
    width: number,
    height: number,
    fps: number,
    pcm: Int16Array,
    sampleRate: number,
    channels: number,
    qp = 26,
  ): Uint8Array {
    const { exports } = this;
    const n = frames.length;
    if (n === 0) throw new RangeError("frames must be non-empty");

    const frameLen = width * height * 3;
    for (let i = 0; i < n; i++) {
      if (frames[i].length !== frameLen) {
        throw new RangeError(
          `frame[${i}] has ${frames[i].length} bytes; expected ${frameLen} (${width}×${height}×3)`,
        );
      }
    }

    const totalLen = frameLen * n;
    const pcmSamples = pcm.length;
    const pcmBytes = pcmSamples * 2; // i16 each

    // --- allocate input video buffer ---
    const inPtr = exports.frffmpeg_alloc(totalLen);
    let mem = new Uint8Array(exports.memory.buffer);
    let offset = inPtr;
    for (let i = 0; i < n; i++) {
      mem.set(frames[i], offset);
      offset += frameLen;
    }

    // --- allocate PCM buffer and write i16 little-endian ---
    // alloc 1 byte minimum even for empty PCM so we have a valid (non-null) pointer.
    const pcmPtr = exports.frffmpeg_alloc(pcmBytes > 0 ? pcmBytes : 1);
    // Re-acquire after alloc; write via DataView for guaranteed LE i16.
    let dv = new DataView(exports.memory.buffer);
    for (let i = 0; i < pcmSamples; i++) {
      dv.setInt16(pcmPtr + i * 2, pcm[i], true);
    }

    // --- allocate 4-byte out_len cell ---
    const outLenPtr = exports.frffmpeg_alloc(4);
    mem = new Uint8Array(exports.memory.buffer);
    mem[outLenPtr] = 0;
    mem[outLenPtr + 1] = 0;
    mem[outLenPtr + 2] = 0;
    mem[outLenPtr + 3] = 0;

    // --- call A/V encoder ---
    const outPtr = exports.frffmpeg_encode_av(
      inPtr,
      frameLen,
      n,
      width,
      height,
      fps,
      qp,
      pcmPtr,
      pcmSamples,
      sampleRate,
      channels,
      outLenPtr,
    );

    // Re-acquire AFTER the encode call: it internally allocs and can grow memory.
    const outLen = new DataView(exports.memory.buffer).getUint32(outLenPtr, true);

    let result: Uint8Array;
    if (outPtr === 0 || outLen === 0) {
      result = new Uint8Array(0);
    } else {
      result = new Uint8Array(exports.memory.buffer, outPtr, outLen).slice();
      exports.frffmpeg_free(outPtr, outLen);
    }

    exports.frffmpeg_free(inPtr, totalLen);
    exports.frffmpeg_free(pcmPtr, pcmBytes > 0 ? pcmBytes : 1);
    exports.frffmpeg_free(outLenPtr, 4);

    return result;
  }

  /**
   * SHARD: encode each frame INDEPENDENTLY into a COMPRESSED AVCC sample (one IDR each).
   * This is the fanout primitive — a frame-range is an independent encode unit, so this
   * can be split across workers/machines. `idr_pic_id` for frame `i` is `startIdrId + i`.
   *
   * @param frames     - Each element is exactly `width * height * 3` bytes of packed RGB24.
   * @param width      - Frame width in pixels.
   * @param height     - Frame height in pixels.
   * @param qp         - Quantiser (0–51; lower = higher quality/bigger). Default 26.
   * @param startIdrId - idr_pic_id of the first frame; subsequent frames increment from it.
   * @returns An array of COPIED Uint8Array samples (one per frame), ready for `muxSamplesToMp4`.
   *
   * Memory-grow safety: re-acquire `memory.buffer` views after every alloc and after each
   * encode call, immediately before each read/write.
   */
  encodeFramesToSamplesCompressed(
    frames: Uint8Array[],
    width: number,
    height: number,
    qp = 26,
    startIdrId = 0,
  ): Uint8Array[] {
    const { exports } = this;
    const n = frames.length;
    if (n === 0) throw new RangeError("frames must be non-empty");

    const frameLen = width * height * 3;
    for (let i = 0; i < n; i++) {
      if (frames[i].length !== frameLen) {
        throw new RangeError(
          `frame[${i}] has ${frames[i].length} bytes; expected ${frameLen} (${width}×${height}×3)`,
        );
      }
    }

    const samples: Uint8Array[] = new Array(n);
    for (let i = 0; i < n; i++) {
      // --- allocate input frame buffer ---
      const inPtr = exports.frffmpeg_alloc(frameLen);
      // Re-acquire mem view: alloc may have grown memory.
      let mem = new Uint8Array(exports.memory.buffer);
      mem.set(frames[i], inPtr);

      // --- allocate 4-byte out_len cell ---
      const outLenPtr = exports.frffmpeg_alloc(4);
      // Re-acquire after second alloc.
      mem = new Uint8Array(exports.memory.buffer);
      mem[outLenPtr] = 0;
      mem[outLenPtr + 1] = 0;
      mem[outLenPtr + 2] = 0;
      mem[outLenPtr + 3] = 0;

      // --- encode this frame independently ---
      const outPtr = exports.frffmpeg_encode_sample_compressed(
        inPtr,
        frameLen,
        width,
        height,
        qp,
        startIdrId + i,
        outLenPtr,
      );

      // Re-acquire AFTER the encode call: it internally allocs and can grow memory.
      const outLen = new DataView(exports.memory.buffer).getUint32(outLenPtr, true);

      if (outPtr === 0 || outLen === 0) {
        samples[i] = new Uint8Array(0);
      } else {
        // Copy before freeing so the caller owns a stable buffer.
        samples[i] = new Uint8Array(exports.memory.buffer, outPtr, outLen).slice();
        exports.frffmpeg_free(outPtr, outLen);
      }

      exports.frffmpeg_free(inPtr, frameLen);
      exports.frffmpeg_free(outLenPtr, 4);
    }

    return samples;
  }

  /**
   * SHARD (rate-controlled): encode `frames` into COMPRESSED AVCC samples (one IDR per frame)
   * whose per-frame QP ADAPTS to hit `targetKbps` (target video bitrate). This is the
   * rate-controlled drop-in for `encodeFramesToSamplesCompressed`: it returns the SAME shape
   * — an array of one COPIED `Uint8Array` sample per input frame, in order — so the worker can
   * swap one call for the other with no other change to the DO fold / mux.
   *
   * Unlike the fixed-QP `encodeFramesToSamplesCompressed` (which is fully per-frame
   * independent and parallelizable), the controller is a SEQUENTIAL feedback loop (frame N's
   * QP depends on frame N−1's measured size), so the whole frame-range is encoded in one wasm
   * call. For the sharded fold, call this once PER SHARD with that shard's contiguous frame
   * range: the controller seeds QP from `targetKbps` and adapts across the shard's own frames.
   *
   * @param frames     - Each element is exactly `width * height * 3` bytes of packed RGB24.
   * @param width      - Frame width in pixels.
   * @param height     - Frame height in pixels.
   * @param fps        - Playback frame rate (sets the per-frame bit budget).
   * @param targetKbps - Target video bitrate in kbit/s.
   * @param startIdrId - idr_pic_id of the first frame (accepted for call-site parity with
   *   `encodeFramesToSamplesCompressed`; every frame is an independent IDR regardless).
   * @returns An array of COPIED Uint8Array samples (one per frame), ready for `muxSamplesToMp4`.
   *
   * Memory-grow safety: re-acquire `memory.buffer` views after every alloc and after the
   * encode call, immediately before each read/write.
   */
  encodeFramesToSamplesCompressedRateControlled(
    frames: Uint8Array[],
    width: number,
    height: number,
    fps: number,
    targetKbps: number,
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    startIdrId = 0,
  ): Uint8Array[] {
    const { exports } = this;
    const n = frames.length;
    if (n === 0) throw new RangeError("frames must be non-empty");

    const frameLen = width * height * 3;
    for (let i = 0; i < n; i++) {
      if (frames[i].length !== frameLen) {
        throw new RangeError(
          `frame[${i}] has ${frames[i].length} bytes; expected ${frameLen} (${width}×${height}×3)`,
        );
      }
    }

    const totalLen = frameLen * n;

    // --- allocate input buffer (all frames concatenated) ---
    const inPtr = exports.frffmpeg_alloc(totalLen);
    // Re-acquire mem view: alloc may have grown memory.
    let mem = new Uint8Array(exports.memory.buffer);
    let offset = inPtr;
    for (let i = 0; i < n; i++) {
      mem.set(frames[i], offset);
      offset += frameLen;
    }

    // --- allocate the OUTPUT u32 lengths array (the call fills it, one per frame) ---
    const lensPtr = exports.frffmpeg_alloc(n * 4);
    // Re-acquire after alloc; zero defensively.
    let dv = new DataView(exports.memory.buffer);
    for (let i = 0; i < n; i++) dv.setUint32(lensPtr + i * 4, 0, true);

    // --- allocate 4-byte out_len cell ---
    const outLenPtr = exports.frffmpeg_alloc(4);
    // Re-acquire after alloc.
    mem = new Uint8Array(exports.memory.buffer);
    mem[outLenPtr] = 0;
    mem[outLenPtr + 1] = 0;
    mem[outLenPtr + 2] = 0;
    mem[outLenPtr + 3] = 0;

    // --- encode the whole range in one call (sequential RC feedback) ---
    const outPtr = exports.frffmpeg_encode_rate_controlled_samples(
      inPtr,
      frameLen,
      n,
      width,
      height,
      fps,
      targetKbps,
      lensPtr,
      outLenPtr,
    );

    // Re-acquire AFTER the encode call: it internally allocs and can grow memory.
    const outLen = new DataView(exports.memory.buffer).getUint32(outLenPtr, true);

    const samples: Uint8Array[] = new Array(n);
    if (outPtr === 0 || outLen === 0) {
      for (let i = 0; i < n; i++) samples[i] = new Uint8Array(0);
    } else {
      // Slice the concatenated blob back into one sample per frame using the lens table,
      // copying each so the caller owns stable buffers — same shape as
      // `encodeFramesToSamplesCompressed`.
      dv = new DataView(exports.memory.buffer);
      let off = outPtr;
      for (let i = 0; i < n; i++) {
        const l = dv.getUint32(lensPtr + i * 4, true);
        samples[i] = new Uint8Array(exports.memory.buffer, off, l).slice();
        off += l;
      }
      exports.frffmpeg_free(outPtr, outLen);
    }

    exports.frffmpeg_free(inPtr, totalLen);
    exports.frffmpeg_free(lensPtr, n * 4);
    exports.frffmpeg_free(outLenPtr, 4);

    return samples;
  }

  /**
   * FOLD: mux pre-encoded shard samples (from `encodeFramesToSamplesCompressed`, in frame
   * order) into a single faststart MP4. Concatenates the samples, builds the u32 length
   * table, and calls the wasm mux.
   *
   * @param samples - AVCC samples in playback order (each is one frame's compressed IDR).
   * @param width   - Frame width in pixels.
   * @param height  - Frame height in pixels.
   * @param fps     - Playback frame rate.
   * @returns A COPIED Uint8Array containing the complete MP4 file.
   *
   * Memory-grow safety: re-acquire `memory.buffer` views after every alloc and after the
   * mux call, immediately before each read/write.
   */
  muxSamplesToMp4(
    samples: Uint8Array[],
    width: number,
    height: number,
    fps: number,
  ): Uint8Array {
    const { exports } = this;
    const n = samples.length;
    if (n === 0) throw new RangeError("samples must be non-empty");

    let totalLen = 0;
    for (let i = 0; i < n; i++) totalLen += samples[i].length;

    // --- allocate concatenated samples buffer ---
    const samplesPtr = exports.frffmpeg_alloc(totalLen);
    // Re-acquire mem view: alloc may have grown memory.
    let mem = new Uint8Array(exports.memory.buffer);
    let off = samplesPtr;
    for (let i = 0; i < n; i++) {
      mem.set(samples[i], off);
      off += samples[i].length;
    }

    // --- allocate the u32 lengths array (4 bytes each) ---
    const lensPtr = exports.frffmpeg_alloc(n * 4);
    // Re-acquire after second alloc.
    let dv = new DataView(exports.memory.buffer);
    for (let i = 0; i < n; i++) {
      dv.setUint32(lensPtr + i * 4, samples[i].length, true);
    }

    // --- allocate 4-byte out_len cell ---
    const outLenPtr = exports.frffmpeg_alloc(4);
    // Re-acquire after third alloc.
    mem = new Uint8Array(exports.memory.buffer);
    mem[outLenPtr] = 0;
    mem[outLenPtr + 1] = 0;
    mem[outLenPtr + 2] = 0;
    mem[outLenPtr + 3] = 0;

    // --- fold ---
    const outPtr = exports.frffmpeg_mux_samples(
      samplesPtr,
      lensPtr,
      n,
      width,
      height,
      fps,
      outLenPtr,
    );

    // Re-acquire AFTER the mux call: it internally allocs and can grow memory.
    const outLen = new DataView(exports.memory.buffer).getUint32(outLenPtr, true);

    let result: Uint8Array;
    if (outPtr === 0 || outLen === 0) {
      result = new Uint8Array(0);
    } else {
      // Copy before freeing so the caller owns a stable buffer.
      result = new Uint8Array(exports.memory.buffer, outPtr, outLen).slice();
      exports.frffmpeg_free(outPtr, outLen);
    }

    exports.frffmpeg_free(samplesPtr, totalLen);
    exports.frffmpeg_free(lensPtr, n * 4);
    exports.frffmpeg_free(outLenPtr, 4);

    return result;
  }

  /**
   * FOLD WITH AUDIO: mux pre-encoded shard video samples (from
   * `encodeFramesToSamplesCompressed`, in frame order) TOGETHER with interleaved i16 PCM
   * (AAC-LC audio) into ONE faststart MP4 with TWO tracks. This is the audio-carrying fold:
   * the SHARDED (fanout) full-length clip path produces only video samples, so the plain
   * `muxSamplesToMp4` gives a video-only MP4 — this method adds sound to sharded clips.
   *
   * @param samples    - AVCC video samples in playback order (each is one frame's IDR).
   * @param width      - Frame width in pixels.
   * @param height     - Frame height in pixels.
   * @param fps        - Playback frame rate.
   * @param pcm        - Interleaved i16 PCM (mono: one value/sample; stereo: L,R,L,R…).
   * @param sampleRate - Audio sample rate in Hz (e.g. 44100).
   * @param channels   - 1 (mono) or 2 (stereo).
   * @returns A COPIED Uint8Array containing the complete two-track MP4 file.
   *
   * Memory-grow safety: re-acquire `memory.buffer` views after every alloc and after the
   * mux call, immediately before each read/write. The PCM is written little-endian i16
   * directly into wasm linear memory.
   */
  muxSamplesToMp4WithAudio(
    samples: Uint8Array[],
    width: number,
    height: number,
    fps: number,
    pcm: Int16Array,
    sampleRate: number,
    channels: number,
  ): Uint8Array {
    const { exports } = this;
    const n = samples.length;
    if (n === 0) throw new RangeError("samples must be non-empty");

    let totalLen = 0;
    for (let i = 0; i < n; i++) totalLen += samples[i].length;

    const pcmSamples = pcm.length;
    const pcmBytes = pcmSamples * 2; // i16 each

    // --- allocate concatenated samples buffer ---
    const samplesPtr = exports.frffmpeg_alloc(totalLen);
    // Re-acquire mem view: alloc may have grown memory.
    let mem = new Uint8Array(exports.memory.buffer);
    let off = samplesPtr;
    for (let i = 0; i < n; i++) {
      mem.set(samples[i], off);
      off += samples[i].length;
    }

    // --- allocate the u32 lengths array (4 bytes each) ---
    const lensPtr = exports.frffmpeg_alloc(n * 4);
    // Re-acquire after alloc.
    let dv = new DataView(exports.memory.buffer);
    for (let i = 0; i < n; i++) {
      dv.setUint32(lensPtr + i * 4, samples[i].length, true);
    }

    // --- allocate PCM buffer and write i16 little-endian ---
    // alloc 1 byte minimum even for empty PCM so we have a valid (non-null) pointer.
    const pcmPtr = exports.frffmpeg_alloc(pcmBytes > 0 ? pcmBytes : 1);
    // Re-acquire after alloc; write via DataView for guaranteed LE i16.
    dv = new DataView(exports.memory.buffer);
    for (let i = 0; i < pcmSamples; i++) {
      dv.setInt16(pcmPtr + i * 2, pcm[i], true);
    }

    // --- allocate 4-byte out_len cell ---
    const outLenPtr = exports.frffmpeg_alloc(4);
    // Re-acquire after alloc.
    mem = new Uint8Array(exports.memory.buffer);
    mem[outLenPtr] = 0;
    mem[outLenPtr + 1] = 0;
    mem[outLenPtr + 2] = 0;
    mem[outLenPtr + 3] = 0;

    // --- fold with audio ---
    const outPtr = exports.frffmpeg_mux_samples_av(
      samplesPtr,
      lensPtr,
      n,
      width,
      height,
      fps,
      pcmPtr,
      pcmSamples,
      sampleRate,
      channels,
      outLenPtr,
    );

    // Re-acquire AFTER the mux call: it internally allocs and can grow memory.
    const outLen = new DataView(exports.memory.buffer).getUint32(outLenPtr, true);

    let result: Uint8Array;
    if (outPtr === 0 || outLen === 0) {
      result = new Uint8Array(0);
    } else {
      // Copy before freeing so the caller owns a stable buffer.
      result = new Uint8Array(exports.memory.buffer, outPtr, outLen).slice();
      exports.frffmpeg_free(outPtr, outLen);
    }

    exports.frffmpeg_free(samplesPtr, totalLen);
    exports.frffmpeg_free(lensPtr, n * 4);
    exports.frffmpeg_free(pcmPtr, pcmBytes > 0 ? pcmBytes : 1);
    exports.frffmpeg_free(outLenPtr, 4);

    return result;
  }
}

// ---------------------------------------------------------------------------
// loadFrffmpeg
// ---------------------------------------------------------------------------

/**
 * Instantiate the frffmpeg WASM module and return an ergonomic `Frffmpeg` wrapper.
 *
 * @param wasmSource - Raw wasm bytes (`BufferSource`) or a pre-compiled
 *   `WebAssembly.Module`. Pass a `Uint8Array` / `ArrayBuffer` in Workers (obtained
 *   via `import wasmUrl from './frffmpeg.wasm'` + `fetch`) or the result of
 *   `WebAssembly.compileStreaming` / a dynamic import.
 *
 * @example Workers usage
 * ```ts
 * import wasmBytes from './frffmpeg.wasm';           // wasm_module binding
 * const enc = await loadFrffmpeg(wasmBytes);
 * const mp4 = enc.encodeRgbFramesToMp4(frames, 640, 480, 30);
 * ```
 *
 * @example Node usage
 * ```ts
 * import { readFileSync } from 'node:fs';
 * const enc = await loadFrffmpeg(readFileSync('./frffmpeg.wasm'));
 * ```
 */
export async function loadFrffmpeg(
  wasmSource: BufferSource | WebAssembly.Module,
): Promise<Frffmpeg> {
  let instance: WebAssembly.Instance;

  if (wasmSource instanceof WebAssembly.Module) {
    instance = await WebAssembly.instantiate(wasmSource, {});
  } else {
    const result = await WebAssembly.instantiate(wasmSource as BufferSource, {});
    instance = result.instance;
  }

  const exports = instance.exports as unknown as FrffmpegExports;

  // Validate that the required exports are present.
  for (const name of ["memory", "frffmpeg_alloc", "frffmpeg_free", "frffmpeg_encode_ipcm", "frffmpeg_encode_compressed", "frffmpeg_encode_rate_controlled", "frffmpeg_encode_rate_controlled_av", "frffmpeg_encode_av", "frffmpeg_encode_sample_compressed", "frffmpeg_encode_rate_controlled_samples", "frffmpeg_mux_samples", "frffmpeg_mux_samples_av"] as const) {
    if (!(name in exports)) {
      throw new Error(`frffmpeg WASM is missing required export: "${name}"`);
    }
  }

  return new Frffmpeg(exports);
}
