// @a0n/gnosis/src/esc/edge-side-compute.ts
/**
 * Edge-Side Compute (ESC) — opt-in, content-addressed memoization for gnosis frontiers.
 *
 * "Computation teleport" is valid only for declared-pure regions: same topology slice,
 * same artifact identity, same canonical input, and fixed runtime ABI must imply the same result.
 *
 * @see aeon-flux ESI (edge-side inference) for the parallel caching pattern.
 * @see `./cache-observation.js` for the Lean `CacheObservation` marginal-fold mirror on ESC paths.
 */

import { alchemicHash48HexSync } from '@a0n/bitwise/cache-fp64';
import { cacheFingerprint48HexSync } from '@a0n/bitwise/cache-fp48';
import {
  getCrossing,
  toBytes,
  type CrossingOp,
  type PackedKnot,
} from '../runtime/packed-knot';

/** Cache policy: only `pure` may participate in cross-peer ESC without extra proofs. */
export type ESCMode = 'none' | 'pure';

/** Optional GG/node metadata shape for future compiler support (documentation contract). */
export interface ESCNodeMetadata {
  readonly esc?: {
    readonly mode: ESCMode;
    /** Extra stable segments folded into {@link computeEscCacheKey} when non-empty. */
    readonly keyParts?: readonly string[];
  };
}

export interface EscPureSegmentHint {
  readonly mode: 'pure';
  /**
   * Packed knot slice whose serialized bytes hash into the cache key.
   * Use {@link slicePackedKnot} on the program knot from `entryCrossingIndex` to the segment end.
   */
  readonly knotSlice: PackedKnot;
  /** Index of the entry crossing in the original full knot (segment identity). */
  readonly entryCrossingIndex: number;
  /** Optional tenant/app namespace; empty means global key (use with care). */
  readonly policySalt?: string;
  /**
   * Positive integer marginal fold-work unit for cold/stale misses (Lean `PositiveFoldWork.val`).
   * Default is 1 in the WASM surface.
   */
  readonly missFoldWork?: number;
  /**
   * When false and the tier returns a matching segment, treat as stale (re-execute, do not serve cache).
   * Models invalidation / phantom coherence per `CacheObservation.lean`.
   */
  readonly cacheEntryValid?: boolean;
}

function packCrossings(ops: CrossingOp[]): Uint8Array {
  const byteCount = Math.ceil(ops.length / 4);
  const bytes = new Uint8Array(byteCount);
  for (let i = 0; i < ops.length; i++) {
    const byteIndex = i >> 2;
    const bitShift = 6 - 2 * (i & 3);
    const op = ops[i] ?? 0;
    bytes[byteIndex] = (bytes[byteIndex] ?? 0) | (op << bitShift);
  }
  return bytes;
}

/**
 * Topological slice [startInclusive, endExclusive) preserving crossing order and node IDs.
 */
export function slicePackedKnot(
  knot: PackedKnot,
  startInclusive: number,
  endExclusive: number
): PackedKnot {
  if (
    startInclusive < 0 ||
    endExclusive > knot.crossingCount ||
    startInclusive >= endExclusive
  ) {
    throw new RangeError(
      `slicePackedKnot: invalid range [${startInclusive}, ${endExclusive}) for crossingCount=${knot.crossingCount}`
    );
  }
  const ops: CrossingOp[] = [];
  const nodeIds: string[] = [];
  for (let i = startInclusive; i < endExclusive; i++) {
    ops.push(getCrossing(knot, i));
    nodeIds.push(knot.nodeIds[i] ?? '_');
  }
  return {
    crossingCount: ops.length,
    crossings: packCrossings(ops),
    nodeIds,
  };
}

/** Stable JSON-like canonicalization for cache keys (sorted object keys, deterministic arrays). */
export function canonicalEscInput(value: unknown): Uint8Array {
  return new TextEncoder().encode(stableStringify(value));
}

function stableStringify(value: unknown): string {
  if (value === null || value === undefined) {
    return 'null';
  }
  if (typeof value === 'string') {
    return JSON.stringify(value);
  }
  if (typeof value === 'number') {
    return Number.isFinite(value) ? String(value) : 'null';
  }
  if (typeof value === 'bigint') {
    return JSON.stringify(value.toString());
  }
  if (typeof value === 'boolean') {
    return value ? 'true' : 'false';
  }
  if (value instanceof Uint8Array) {
    return `"u8:${alchemicHash48HexSync(value)}"`;
  }
  if (Array.isArray(value)) {
    return `[${value.map((item) => stableStringify(item)).join(',')}]`;
  }
  if (typeof value === 'object') {
    const entries = Object.entries(value)
      .filter(([, v]) => {
        const t = typeof v;
        return t !== 'undefined' && t !== 'function' && t !== 'symbol';
      })
      .sort(([a], [b]) => a.localeCompare(b));
    return `{${entries
      .map(([k, v]) => `${JSON.stringify(k)}:${stableStringify(v)}`)
      .join(',')}}`;
  }
  return 'null';
}

export interface ComputeEscCacheKeyInput {
  readonly knotSlice: PackedKnot;
  readonly entryCrossingIndex: number;
  readonly canonicalInput: Uint8Array;
  readonly artifactFingerprint: string;
  readonly runtimeBindingHash?: string | null;
  readonly policySalt?: string;
  /** Optional extra parts (e.g. export name, ABI tag). */
  readonly handlerVersionParts?: readonly string[];
}

/** Version prefix — bump when key material or digest algorithm changes. */
const ESC_KEY_PREFIX_FP64 = new TextEncoder().encode('gnosis:esc:fp64-v1\0');

/**
 * Opt-in fp48 version prefix. Coexists with `gnosis:esc:fp64-v1`; selected
 * only when `ESC_FP48_ENABLED` is truthy (see {@link isFp48Enabled}).
 *
 * Provenance: `BizarroCompiler.fp48Compressor` + `BizarroCompiler.bitwise_shrink_4_3`
 * + `BizarroCompiler.fp48_concrete_shrink` (`48 * 4 = 64 * 3`) in
 * `open-source/gnosis/lean/Lean/ForkRaceFoldTheorems/BizarroCompiler.lean`.
 */
const ESC_KEY_PREFIX_FP48 = new TextEncoder().encode('gnosis:esc:fp48-v1\0');

/** Name of the environment flag that switches ESC keys to the fp48 digest. */
export const ESC_FP48_ENV_FLAG = 'ESC_FP48_ENABLED' as const;

/**
 * Read `ESC_FP48_ENABLED` from the ambient environment (Node / Deno / Bun).
 * Truthy values: `'1'`, `'true'`, `'yes'`, `'on'` (case-insensitive). Anything
 * else — including unset — leaves fp48 disabled, so flag-off is bit-identical
 * to the pre-change baseline.
 */
function isFp48Enabled(): boolean {
  const proc: { env?: Record<string, string | undefined> } | undefined = (
    globalThis as { process?: { env?: Record<string, string | undefined> } }
  ).process;
  const raw = proc?.env?.[ESC_FP48_ENV_FLAG];
  if (typeof raw !== 'string') {
    return false;
  }
  const v = raw.trim().toLowerCase();
  return v === '1' || v === 'true' || v === 'yes' || v === 'on';
}

/**
 * Content-addressed ESC key. Two coexisting schemes:
 *
 * - `gnosis:esc:fp64-v1` (default): digest via `alchemicHash48HexSync` from
 *   `@a0n/bitwise/cache-fp64`; matches the pre-fp48 baseline bit-for-bit.
 * - `gnosis:esc:fp48-v1` (opt-in, `ESC_FP48_ENABLED=1`): digest via
 *   `cacheFingerprint48HexSync` from `@a0n/bitwise/cache-fp48` — 12 lowercase
 *   hex chars. The 4:3 shrink is the concrete witness of
 *   `BizarroCompiler.fp48Compressor` (`outputBits * 4 = inputBits * 3`), see
 *   `BizarroCompiler.fp48_concrete_shrink` for the `48 * 4 = 64 * 3` instance.
 *
 * Both paths are stable for identical inputs; downstream consumers MUST still
 * verify artifact/export/abi on hit (48-bit space offers ~2²⁴ birthday
 * resistance — cache coherence, not crypto).
 */
export function computeEscCacheKey(input: ComputeEscCacheKeyInput): string {
  const fp48 = isFp48Enabled();
  const prefix = fp48 ? ESC_KEY_PREFIX_FP48 : ESC_KEY_PREFIX_FP64;
  const knotBytes = toBytes(input.knotSlice);
  const idx = new Uint8Array(4);
  new DataView(idx.buffer).setUint32(0, input.entryCrossingIndex >>> 0, false);

  const parts: Uint8Array[] = [
    prefix,
    idx,
    knotBytes,
    input.canonicalInput,
    new TextEncoder().encode(input.artifactFingerprint),
    new TextEncoder().encode(input.runtimeBindingHash ?? ''),
    new TextEncoder().encode(input.policySalt ?? ''),
  ];
  if (input.handlerVersionParts?.length) {
    for (const p of input.handlerVersionParts) {
      parts.push(new TextEncoder().encode(`\0${p}`));
    }
  }

  const total = parts.reduce((s, b) => s + b.length, 0);
  const concat = new Uint8Array(total);
  let o = 0;
  for (const b of parts) {
    concat.set(b, o);
    o += b.length;
  }

  return fp48 ? cacheFingerprint48HexSync(concat) : alchemicHash48HexSync(concat);
}

/** Serializable WASM ESC segment (teleport payload). */
export interface EscWasmSegmentRecord {
  readonly version: 1;
  readonly artifactId: string;
  readonly exportName: string;
  readonly abi: string;
  readonly payload: unknown;
  readonly logs: string;
}

/**
 * Handles the gnosis encode Esc Wasm Segment workflow.
 */
export function encodeEscWasmSegment(record: EscWasmSegmentRecord): Uint8Array {
  return new TextEncoder().encode(
    JSON.stringify({
      version: record.version,
      artifactId: record.artifactId,
      exportName: record.exportName,
      abi: record.abi,
      payload: record.payload,
      logs: record.logs,
    })
  );
}

/**
 * Handles the gnosis decode Esc Wasm Segment workflow.
 */
export function decodeEscWasmSegment(bytes: Uint8Array): EscWasmSegmentRecord {
  const text = new TextDecoder('utf-8', { fatal: true }).decode(bytes);
  const parsed: unknown = JSON.parse(text);
  if (
    typeof parsed !== 'object' ||
    parsed === null ||
    (parsed as { version?: unknown }).version !== 1
  ) {
    throw new Error('decodeEscWasmSegment: invalid record');
  }
  const r = parsed as Record<string, unknown>;
  return {
    version: 1,
    artifactId: String(r['artifactId'] ?? ''),
    exportName: String(r['exportName'] ?? ''),
    abi: String(r['abi'] ?? ''),
    payload: r['payload'],
    logs: String(r['logs'] ?? ''),
  };
}

export interface EscTier {
  get(key: string): Promise<EscWasmSegmentRecord | null>;
  set(key: string, record: EscWasmSegmentRecord): Promise<void>;
}

/** In-memory ESC tier (process-local). */
export class EscMemoryTier implements EscTier {
  private readonly entries = new Map<string, EscWasmSegmentRecord>();

  async get(key: string): Promise<EscWasmSegmentRecord | null> {
    return this.entries.get(key) ?? null;
  }

  async set(key: string, record: EscWasmSegmentRecord): Promise<void> {
    this.entries.set(key, record);
  }

  clear(): void {
    this.entries.clear();
  }

  get size(): number {
    return this.entries.size;
  }
}
