// @a0n/gnosis/src/auth/bootstrap.ts
import type { ASTEdge, ASTNode, GraphAST } from '../betty/compiler';
import { registerCoreRuntimeHandlers } from '../runtime/core-handlers';
import { GnosisRegistry } from '../runtime/registry';
import type { GnosisHandler } from '../runtime/registry';
import {
  mergeExecutionAuthContexts,
  normalizeExecutionAuthContext,
  type GnosisExecutionAuthContext,
} from './core';
import {
  GNOSIS_CORE_AUTH_LABELS,
  registerCoreAuthHandlers,
} from './handlers';

const SAFE_BOOTSTRAP_LABELS = new Set<string>([
  ...Object.values(GNOSIS_CORE_AUTH_LABELS),
  'Result',
  'Option',
  'Variant',
  'Destructure',
]);

const DEFAULT_BOOTSTRAP_MAX_STEPS = 32;

export interface GnosisExecutionAuthBootstrapOptions {
  initialPayload?: unknown;
  maxSteps?: number;
}

export interface GnosisExecutionAuthBootstrapResult {
  executionAuth: GnosisExecutionAuthContext | null;
  attempted: boolean;
  visitedNodeIds: string[];
}

function rootNodeIds(ast: GraphAST): string[] {
  const targetIds = new Set<string>();
  for (const edge of ast.edges) {
    for (const targetId of edge.targetIds) {
      targetIds.add(targetId.trim());
    }
  }

  return [...ast.nodes.keys()].filter(
    (nodeId) => !targetIds.has(nodeId.trim())
  );
}

function findBootstrapHandler(
  registry: GnosisRegistry,
  node: ASTNode
): GnosisHandler | null {
  for (const label of node.labels) {
    if (!SAFE_BOOTSTRAP_LABELS.has(label)) {
      continue;
    }

    const handler = registry.getHandler(label);
    if (handler) {
      return handler;
    }
  }

  return null;
}

function nextBootstrapEdge(ast: GraphAST, sourceId: string): ASTEdge | null {
  const outgoingEdges = ast.edges.filter((edge) =>
    edge.sourceIds.some((candidate) => candidate.trim() === sourceId)
  );
  if (outgoingEdges.length !== 1) {
    return null;
  }

  const [edge] = outgoingEdges;
  if (!edge || edge.type !== 'PROCESS' || edge.targetIds.length !== 1) {
    return null;
  }

  return edge;
}

/**
 * Handles the gnosis bootstrap Execution Auth From Topology workflow.
 */
export async function bootstrapExecutionAuthFromTopology(
  ast: GraphAST,
  options: GnosisExecutionAuthBootstrapOptions = {}
): Promise<GnosisExecutionAuthBootstrapResult> {
  const registry = new GnosisRegistry();
  registerCoreAuthHandlers(registry);
  registerCoreRuntimeHandlers(registry);

  const maxSteps = options.maxSteps ?? DEFAULT_BOOTSTRAP_MAX_STEPS;
  const initialPayload = options.initialPayload ?? null;
  const visitedNodeIds: string[] = [];
  let attempted = false;
  let mergedExecutionAuth =
    normalizeExecutionAuthContext(
      typeof initialPayload === 'object' && initialPayload !== null
        ? (initialPayload as Record<string, unknown>).executionAuth
        : null
    ) ?? null;

  for (const rootNodeId of rootNodeIds(ast)) {
    let currentNodeId: string | null = rootNodeId.trim();
    let currentPayload = initialPayload;
    let currentExecutionAuth = mergedExecutionAuth;
    let steps = 0;

    while (currentNodeId && steps < maxSteps) {
      const node = ast.nodes.get(currentNodeId);
      if (!node) {
        break;
      }

      const handler = findBootstrapHandler(registry, node);
      if (!handler) {
        break;
      }

      attempted = true;
      visitedNodeIds.push(currentNodeId);
      const handlerContext = {
        nodeId: currentNodeId,
        ...(currentExecutionAuth !== undefined && currentExecutionAuth !== null
          ? { executionAuth: currentExecutionAuth }
          : {}),
      };
      currentPayload = await handler(
        currentPayload,
        node.properties,
        handlerContext
      );

      const nextExecutionAuth = normalizeExecutionAuthContext(
        typeof currentPayload === 'object' && currentPayload !== null
          ? (currentPayload as Record<string, unknown>).executionAuth
          : null
      );
      if (nextExecutionAuth) {
        currentExecutionAuth = currentExecutionAuth
          ? mergeExecutionAuthContexts(currentExecutionAuth, nextExecutionAuth)
          : nextExecutionAuth;
        mergedExecutionAuth = mergedExecutionAuth
          ? mergeExecutionAuthContexts(
              mergedExecutionAuth,
              currentExecutionAuth
            )
          : currentExecutionAuth;
      }

      const edge = nextBootstrapEdge(ast, currentNodeId);
      if (!edge) {
        break;
      }

      currentNodeId = edge.targetIds[0]?.trim() ?? null;
      steps += 1;
    }
  }

  return {
    executionAuth: mergedExecutionAuth,
    attempted,
    visitedNodeIds,
  };
}
