// @a0n/gnosis/src/auth/encoding.ts
interface BufferLike extends Uint8Array {
  toString(encoding?: string): string;
}

interface BufferLikeConstructor {
  from(input: Uint8Array | string, encoding?: string): BufferLike;
}

function getBufferConstructor(): BufferLikeConstructor | null {
  const candidate = (globalThis as { Buffer?: BufferLikeConstructor }).Buffer;
  return candidate && typeof candidate.from === 'function' ? candidate : null;
}

function bytesToBinaryString(bytes: Uint8Array): string {
  const chunks: string[] = [];
  const chunkSize = 0x8000;

  for (let index = 0; index < bytes.length; index += chunkSize) {
    const chunk = bytes.subarray(index, index + chunkSize);
    chunks.push(String.fromCharCode(...chunk));
  }

  return chunks.join('');
}

function binaryStringToBytes(binary: string): Uint8Array {
  const bytes = new Uint8Array(binary.length);
  for (let index = 0; index < binary.length; index += 1) {
    bytes[index] = binary.charCodeAt(index);
  }
  return bytes;
}

/**
 * Handles the gnosis encode Base64 workflow.
 */
export function encodeBase64(bytes: Uint8Array | ArrayBuffer): string {
  const source = bytes instanceof Uint8Array ? bytes : new Uint8Array(bytes);

  if (typeof btoa === 'function') {
    return btoa(bytesToBinaryString(source));
  }

  const BufferCtor = getBufferConstructor();
  if (BufferCtor) {
    return BufferCtor.from(source).toString('base64');
  }

  throw new Error('Base64 encoder unavailable in current runtime.');
}

/**
 * Handles the gnosis decode Base64 workflow.
 */
export function decodeBase64(value: string): Uint8Array {
  const normalized = value.replace(/\s+/g, '');

  if (typeof atob === 'function') {
    return binaryStringToBytes(atob(normalized));
  }

  const BufferCtor = getBufferConstructor();
  if (BufferCtor) {
    return new Uint8Array(BufferCtor.from(normalized, 'base64'));
  }

  throw new Error('Base64 decoder unavailable in current runtime.');
}

/**
 * Handles the gnosis encode Base64 Url workflow.
 */
export function encodeBase64Url(bytes: Uint8Array | ArrayBuffer): string {
  return encodeBase64(bytes)
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/=+$/u, '');
}

/**
 * Handles the gnosis decode Base64 Url workflow.
 */
export function decodeBase64Url(value: string): Uint8Array {
  const normalized = value.replace(/-/g, '+').replace(/_/g, '/');
  const padded = normalized + '='.repeat((4 - (normalized.length % 4)) % 4);
  return decodeBase64(padded);
}

/**
 * Handles the gnosis encode Hex workflow.
 */
export function encodeHex(bytes: Uint8Array): string {
  return Array.from(bytes, (byte) => byte.toString(16).padStart(2, '0')).join(
    ''
  );
}

/**
 * Handles the gnosis decode Hex workflow.
 */
export function decodeHex(value: string): Uint8Array {
  const normalized = value.trim();
  if (normalized.length === 0 || normalized.length % 2 !== 0) {
    throw new Error('Hex payload must contain an even number of characters.');
  }

  if (!/^[0-9a-fA-F]+$/u.test(normalized)) {
    throw new Error('Hex payload contains non-hex characters.');
  }

  const bytes = new Uint8Array(normalized.length / 2);
  for (let index = 0; index < normalized.length; index += 2) {
    bytes[index / 2] = Number.parseInt(normalized.slice(index, index + 2), 16);
  }
  return bytes;
}
