// @a0n/gnosis/src/auth/tee-attestation.ts
import { decodeBase64Url, encodeBase64Url } from './encoding';

const HALT_ATTESTATION_ALGORITHM = 'ES256' as const;
const ECDSA_SIGNING_ALGORITHM: EcKeyImportParams = {
  name: 'ECDSA',
  namedCurve: 'P-256',
};
const ECDSA_SIGNATURE_ALGORITHM: EcdsaParams = {
  name: 'ECDSA',
  hash: 'SHA-256',
};

export type HaltAttestationAlgorithm = typeof HALT_ATTESTATION_ALGORITHM;

export interface HaltAttestationClaims {
  measurement: string;
  programHash: string;
  vkHash: string;
  publicSignalsHash: string;
  nonce: string;
  issuedAt: number;
  expiresAt: number;
  tcbVersion?: string;
}

export interface HaltAttestationEnvelope {
  algorithm: HaltAttestationAlgorithm;
  signature: string;
  claims: HaltAttestationClaims;
  keyId?: string;
  publicKey?: JsonWebKey;
}

export interface HaltExecutionEnvelope {
  proof: string;
  publicSignals: unknown;
  publicSignalsHash: string;
  programHash: string;
  vkHash: string;
  nonce: string;
  expiresAt: number;
  attestation: HaltAttestationEnvelope;
}

export interface CreateHaltAttestationOptions {
  claims: HaltAttestationClaims;
  privateKey: JsonWebKey;
  keyId?: string;
  publicKey?: JsonWebKey;
}

export interface HaltAttestationVerificationOptions {
  nowMs?: number;
  maxAttestationAgeMs?: number;
  trustedMeasurements?: readonly string[];
  trustedProgramHashes?: readonly string[];
  trustedVkHashes?: readonly string[];
  trustedPublicKeys?: readonly JsonWebKey[];
  trustedKeyById?: Readonly<Record<string, JsonWebKey>>;
  allowInlinePublicKey?: boolean;
  minTcbVersion?: string;
  expectedProgramHash?: string;
  expectedVkHash?: string;
  expectedPublicSignalsHash?: string;
  expectedNonce?: string;
  expectedExpiresAt?: number;
  nonceStore?: NonceReplayStore;
}

export interface HaltAttestationVerificationResult {
  valid: boolean;
  reason?: string;
  claims?: HaltAttestationClaims;
}

export interface ZkProofVerifierInput {
  proof: string;
  publicSignals: unknown;
  publicSignalsHash: string;
  programHash: string;
  vkHash: string;
}

export type ZkProofVerifier = (
  input: ZkProofVerifierInput
) => Promise<boolean> | boolean;

export interface ZkExecutionVerificationOptions
  extends HaltAttestationVerificationOptions {
  proofVerifier?: ZkProofVerifier;
  proofVerifiedHint?: boolean;
  requireProofVerification?: boolean;
}

export interface ZkExecutionVerificationResult {
  valid: boolean;
  reason?: string;
  proofVerified: boolean;
  attestation: HaltAttestationVerificationResult;
}

export interface NonceReplayStore {
  has(nonce: string, nowMs: number): boolean;
  mark(nonce: string, expiresAt: number): void;
  prune(nowMs: number): void;
}

export class InMemoryNonceReplayStore implements NonceReplayStore {
  private readonly expirations = new Map<string, number>();

  public has(nonce: string, nowMs: number): boolean {
    this.prune(nowMs);
    const expiration = this.expirations.get(nonce);
    return typeof expiration === 'number' && expiration > nowMs;
  }

  public mark(nonce: string, expiresAt: number): void {
    this.expirations.set(nonce, expiresAt);
  }

  public prune(nowMs: number): void {
    for (const [nonce, expiration] of this.expirations.entries()) {
      if (expiration <= nowMs) {
        this.expirations.delete(nonce);
      }
    }
  }
}

function isObjectRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function isNonEmptyString(value: unknown): value is string {
  return typeof value === 'string' && value.trim().length > 0;
}

function normalizeList(values: readonly string[] | undefined): Set<string> {
  if (!values || values.length === 0) {
    return new Set();
  }

  const normalized = values
    .map((value) => value.trim())
    .filter((value) => value.length > 0);

  return new Set(normalized);
}

function normalizeVersion(version: string): number[] {
  return version
    .split('.')
    .map((piece) => {
      const parsed = Number.parseInt(piece, 10);
      return Number.isFinite(parsed) ? parsed : 0;
    })
    .slice(0, 4);
}

function compareVersion(left: string, right: string): number {
  const leftParts = normalizeVersion(left);
  const rightParts = normalizeVersion(right);
  const maxLength = Math.max(leftParts.length, rightParts.length);

  for (let index = 0; index < maxLength; index += 1) {
    const leftValue = leftParts[index] ?? 0;
    const rightValue = rightParts[index] ?? 0;
    if (leftValue > rightValue) return 1;
    if (leftValue < rightValue) return -1;
  }

  return 0;
}

function stableStringify(value: unknown): string {
  if (value === null) {
    return 'null';
  }

  if (value === undefined) {
    return 'null';
  }

  if (typeof value === 'string') {
    return JSON.stringify(value);
  }

  if (typeof value === 'number') {
    return Number.isFinite(value) ? String(value) : 'null';
  }

  if (typeof value === 'boolean') {
    return value ? 'true' : 'false';
  }

  if (typeof value === 'bigint') {
    return JSON.stringify(value.toString(10));
  }

  if (Array.isArray(value)) {
    return `[${value.map((item) => stableStringify(item)).join(',')}]`;
  }

  if (typeof value === 'object') {
    const record = value as Record<string, unknown>;
    const entries = Object.entries(record)
      .filter(([, entryValue]) => {
        const entryType = typeof entryValue;
        return (
          entryType !== 'undefined' &&
          entryType !== 'function' &&
          entryType !== 'symbol'
        );
      })
      .sort(([leftKey], [rightKey]) => leftKey.localeCompare(rightKey));

    return `{${entries
      .map(
        ([key, entryValue]) =>
          `${JSON.stringify(key)}:${stableStringify(entryValue)}`
      )
      .join(',')}}`;
  }

  return 'null';
}

function toBase64Url(bytes: Uint8Array | ArrayBuffer): string {
  return encodeBase64Url(bytes);
}

function fromBase64Url(value: string): Uint8Array {
  return decodeBase64Url(value);
}

function toArrayBuffer(bytes: Uint8Array): ArrayBuffer {
  return bytes.buffer.slice(
    bytes.byteOffset,
    bytes.byteOffset + bytes.byteLength
  ) as ArrayBuffer;
}

async function sha256Hex(input: string): Promise<string> {
  const digest = await crypto.subtle.digest(
    'SHA-256',
    new TextEncoder().encode(input)
  );
  return Array.from(new Uint8Array(digest), (byte) =>
    byte.toString(16).padStart(2, '0')
  ).join('');
}

function coerceClaims(value: unknown): HaltAttestationClaims | null {
  const record = isObjectRecord(value) ? value : null;
  if (!record) {
    return null;
  }

  const measurement = record.measurement;
  const programHash = record.programHash;
  const vkHash = record.vkHash;
  const publicSignalsHash = record.publicSignalsHash;
  const nonce = record.nonce;
  const issuedAt = record.issuedAt;
  const expiresAt = record.expiresAt;
  const tcbVersion = record.tcbVersion;

  if (
    !isNonEmptyString(measurement) ||
    !isNonEmptyString(programHash) ||
    !isNonEmptyString(vkHash) ||
    !isNonEmptyString(publicSignalsHash) ||
    !isNonEmptyString(nonce) ||
    typeof issuedAt !== 'number' ||
    !Number.isFinite(issuedAt) ||
    typeof expiresAt !== 'number' ||
    !Number.isFinite(expiresAt)
  ) {
    return null;
  }

  if (typeof tcbVersion !== 'undefined' && !isNonEmptyString(tcbVersion)) {
    return null;
  }

  return {
    measurement: measurement.trim(),
    programHash: programHash.trim(),
    vkHash: vkHash.trim(),
    publicSignalsHash: publicSignalsHash.trim(),
    nonce: nonce.trim(),
    issuedAt,
    expiresAt,
    ...(isNonEmptyString(tcbVersion) && { tcbVersion: tcbVersion.trim() }),
  };
}

function signaturePayload(claims: HaltAttestationClaims): Uint8Array {
  const serialized = stableStringify(claims);
  return new TextEncoder().encode(serialized);
}

async function importPublicKey(key: JsonWebKey): Promise<CryptoKey> {
  return crypto.subtle.importKey('jwk', key, ECDSA_SIGNING_ALGORITHM, false, [
    'verify',
  ]);
}

async function importPrivateKey(key: JsonWebKey): Promise<CryptoKey> {
  return crypto.subtle.importKey('jwk', key, ECDSA_SIGNING_ALGORITHM, false, [
    'sign',
  ]);
}

function resolveTrustedPublicKeys(
  attestation: HaltAttestationEnvelope,
  options: HaltAttestationVerificationOptions
): JsonWebKey[] {
  const resolved: JsonWebKey[] = [];

  if (attestation.keyId && options.trustedKeyById) {
    const key = options.trustedKeyById[attestation.keyId];
    if (key) {
      resolved.push(key);
    }
  }

  if (options.trustedPublicKeys && options.trustedPublicKeys.length > 0) {
    resolved.push(...options.trustedPublicKeys);
  }

  if (options.allowInlinePublicKey === true && attestation.publicKey) {
    resolved.push(attestation.publicKey);
  }

  const deduped = new Map<string, JsonWebKey>();
  for (const key of resolved) {
    deduped.set(stableStringify(key), key);
  }

  return Array.from(deduped.values());
}

async function verifyAttestationSignature(
  attestation: HaltAttestationEnvelope,
  candidateKeys: readonly JsonWebKey[]
): Promise<boolean> {
  if (candidateKeys.length === 0) {
    return false;
  }

  const signature = fromBase64Url(attestation.signature);
  const payload = toArrayBuffer(signaturePayload(attestation.claims));
  const signatureBytes = toArrayBuffer(signature);

  for (const candidateKey of candidateKeys) {
    try {
      const key = await importPublicKey(candidateKey);
      const verified = await crypto.subtle.verify(
        ECDSA_SIGNATURE_ALGORITHM,
        key,
        signatureBytes,
        payload
      );
      if (verified) {
        return true;
      }
    } catch {
      // Continue checking additional trusted keys.
    }
  }

  return false;
}

function checkExpectedValue(
  actual: string,
  expected: string | undefined,
  label: string
): HaltAttestationVerificationResult | null {
  if (!expected) {
    return null;
  }

  if (actual !== expected) {
    return {
      valid: false,
      reason: `${label} mismatch in attestation claims.`,
    };
  }

  return null;
}

/**
 * Returns whether hash Public Signals is true.
 */
export async function hashPublicSignals(
  publicSignals: unknown
): Promise<string> {
  return sha256Hex(stableStringify(publicSignals));
}

/**
 * Handles the gnosis as Halt Attestation Envelope workflow.
 */
export function asHaltAttestationEnvelope(
  value: unknown
): HaltAttestationEnvelope | null {
  const record = isObjectRecord(value) ? value : null;
  if (!record) {
    return null;
  }

  const algorithm = record.algorithm;
  const signature = record.signature;
  const keyId = record.keyId;
  const publicKey = record.publicKey;
  const claims = coerceClaims(record.claims);

  if (algorithm !== HALT_ATTESTATION_ALGORITHM) {
    return null;
  }

  if (!isNonEmptyString(signature) || !claims) {
    return null;
  }

  if (typeof keyId !== 'undefined' && !isNonEmptyString(keyId)) {
    return null;
  }

  if (typeof publicKey !== 'undefined' && !isObjectRecord(publicKey)) {
    return null;
  }

  return {
    algorithm: HALT_ATTESTATION_ALGORITHM,
    signature,
    claims,
    ...(isNonEmptyString(keyId) && { keyId }),
    ...(isObjectRecord(publicKey) && {
      publicKey: publicKey as unknown as JsonWebKey,
    }),
  };
}

/**
 * Handles the gnosis as Halt Execution Envelope workflow.
 */
export function asHaltExecutionEnvelope(
  value: unknown
): HaltExecutionEnvelope | null {
  const record = isObjectRecord(value) ? value : null;
  if (!record) {
    return null;
  }

  const proof = record.proof;
  const publicSignalsHash = record.publicSignalsHash;
  const programHash = record.programHash;
  const vkHash = record.vkHash;
  const nonce = record.nonce;
  const expiresAt = record.expiresAt;
  const attestation = asHaltAttestationEnvelope(record.attestation);

  if (
    !isNonEmptyString(proof) ||
    !isNonEmptyString(publicSignalsHash) ||
    !isNonEmptyString(programHash) ||
    !isNonEmptyString(vkHash) ||
    !isNonEmptyString(nonce) ||
    typeof expiresAt !== 'number' ||
    !Number.isFinite(expiresAt) ||
    !attestation
  ) {
    return null;
  }

  return {
    proof,
    publicSignals: record.publicSignals,
    publicSignalsHash,
    programHash,
    vkHash,
    nonce,
    expiresAt,
    attestation,
  };
}

/**
 * Creates the Halt Attestation.
 */
export async function createHaltAttestation(
  options: CreateHaltAttestationOptions
): Promise<HaltAttestationEnvelope> {
  const claims = coerceClaims(options.claims);
  if (!claims) {
    throw new Error('Invalid HALT attestation claims.');
  }

  if (claims.expiresAt <= claims.issuedAt) {
    throw new Error(
      'HALT attestation expiresAt must be greater than issuedAt.'
    );
  }

  const key = await importPrivateKey(options.privateKey);
  const payload = toArrayBuffer(signaturePayload(claims));
  const signatureBuffer = await crypto.subtle.sign(
    ECDSA_SIGNATURE_ALGORITHM,
    key,
    payload
  );

  return {
    algorithm: HALT_ATTESTATION_ALGORITHM,
    signature: toBase64Url(signatureBuffer),
    claims,
    ...(isNonEmptyString(options.keyId) && { keyId: options.keyId }),
    ...(options.publicKey !== undefined && { publicKey: options.publicKey }),
  };
}

/**
 * Handles the gnosis verify Halt Attestation workflow.
 */
export async function verifyHaltAttestation(
  attestation: HaltAttestationEnvelope,
  options: HaltAttestationVerificationOptions = {}
): Promise<HaltAttestationVerificationResult> {
  const nowMs = options.nowMs ?? Date.now();
  const trustedMeasurements = normalizeList(options.trustedMeasurements);
  const trustedProgramHashes = normalizeList(options.trustedProgramHashes);
  const trustedVkHashes = normalizeList(options.trustedVkHashes);

  const claims = coerceClaims(attestation.claims);
  if (!claims) {
    return { valid: false, reason: 'Invalid attestation claims payload.' };
  }

  if (claims.expiresAt <= claims.issuedAt) {
    return {
      valid: false,
      reason: 'Attestation has invalid issuedAt/expiresAt bounds.',
    };
  }

  if (claims.expiresAt <= nowMs) {
    return { valid: false, reason: 'Attestation is expired.' };
  }

  if (claims.issuedAt > nowMs) {
    return { valid: false, reason: 'Attestation issuedAt is in the future.' };
  }

  if (
    typeof options.maxAttestationAgeMs === 'number' &&
    Number.isFinite(options.maxAttestationAgeMs) &&
    nowMs - claims.issuedAt > options.maxAttestationAgeMs
  ) {
    return {
      valid: false,
      reason: 'Attestation age exceeds maxAttestationAgeMs policy.',
    };
  }

  const expectedProgramMismatch = checkExpectedValue(
    claims.programHash,
    options.expectedProgramHash,
    'programHash'
  );
  if (expectedProgramMismatch) {
    return expectedProgramMismatch;
  }

  const expectedVkMismatch = checkExpectedValue(
    claims.vkHash,
    options.expectedVkHash,
    'vkHash'
  );
  if (expectedVkMismatch) {
    return expectedVkMismatch;
  }

  const expectedSignalsMismatch = checkExpectedValue(
    claims.publicSignalsHash,
    options.expectedPublicSignalsHash,
    'publicSignalsHash'
  );
  if (expectedSignalsMismatch) {
    return expectedSignalsMismatch;
  }

  const expectedNonceMismatch = checkExpectedValue(
    claims.nonce,
    options.expectedNonce,
    'nonce'
  );
  if (expectedNonceMismatch) {
    return expectedNonceMismatch;
  }

  if (
    typeof options.expectedExpiresAt === 'number' &&
    Number.isFinite(options.expectedExpiresAt) &&
    claims.expiresAt !== options.expectedExpiresAt
  ) {
    return {
      valid: false,
      reason: 'expiresAt mismatch in attestation claims.',
    };
  }

  if (
    trustedMeasurements.size > 0 &&
    !trustedMeasurements.has(claims.measurement)
  ) {
    return {
      valid: false,
      reason: `Measurement ${claims.measurement} is not allowlisted.`,
    };
  }

  if (
    trustedProgramHashes.size > 0 &&
    !trustedProgramHashes.has(claims.programHash)
  ) {
    return {
      valid: false,
      reason: 'Program hash is not allowlisted.',
    };
  }

  if (trustedVkHashes.size > 0 && !trustedVkHashes.has(claims.vkHash)) {
    return {
      valid: false,
      reason: 'Verifier key hash is not allowlisted.',
    };
  }

  if (options.minTcbVersion) {
    if (!claims.tcbVersion) {
      return {
        valid: false,
        reason: 'Attestation missing tcbVersion required by policy.',
      };
    }

    if (compareVersion(claims.tcbVersion, options.minTcbVersion) < 0) {
      return {
        valid: false,
        reason: `Attestation tcbVersion ${claims.tcbVersion} below required ${options.minTcbVersion}.`,
      };
    }
  }

  if (options.nonceStore) {
    if (options.nonceStore.has(claims.nonce, nowMs)) {
      return {
        valid: false,
        reason: 'Nonce replay detected for attestation envelope.',
      };
    }
  }

  const trustedPublicKeys = resolveTrustedPublicKeys(attestation, options);
  if (trustedPublicKeys.length === 0) {
    return {
      valid: false,
      reason:
        'No trusted public keys resolved. Provide trustedKeyById/trustedPublicKeys or allow inline key explicitly.',
    };
  }

  const signatureValid = await verifyAttestationSignature(
    attestation,
    trustedPublicKeys
  );
  if (!signatureValid) {
    return {
      valid: false,
      reason: 'HALT attestation signature verification failed.',
    };
  }

  if (options.nonceStore) {
    options.nonceStore.mark(claims.nonce, claims.expiresAt);
  }

  return {
    valid: true,
    claims,
  };
}

/**
 * Handles the gnosis verify Zk Execution Envelope workflow.
 */
export async function verifyZkExecutionEnvelope(
  envelope: HaltExecutionEnvelope,
  options: ZkExecutionVerificationOptions = {}
): Promise<ZkExecutionVerificationResult> {
  const nowMs = options.nowMs ?? Date.now();
  const computedSignalsHash = await hashPublicSignals(envelope.publicSignals);

  if (computedSignalsHash !== envelope.publicSignalsHash) {
    return {
      valid: false,
      reason: 'publicSignalsHash does not match canonicalized publicSignals.',
      proofVerified: false,
      attestation: {
        valid: false,
        reason: 'Execution envelope publicSignalsHash mismatch.',
      },
    };
  }

  if (envelope.expiresAt <= nowMs) {
    return {
      valid: false,
      reason: 'Execution envelope is expired.',
      proofVerified: false,
      attestation: {
        valid: false,
        reason: 'Execution envelope expired before verification.',
      },
    };
  }

  const attestationResult = await verifyHaltAttestation(envelope.attestation, {
    ...options,
    expectedProgramHash: envelope.programHash,
    expectedVkHash: envelope.vkHash,
    expectedPublicSignalsHash: envelope.publicSignalsHash,
    expectedNonce: envelope.nonce,
    expectedExpiresAt: envelope.expiresAt,
  });

  if (!attestationResult.valid) {
    return {
      valid: false,
      ...(attestationResult.reason !== undefined && {
        reason: attestationResult.reason,
      }),
      proofVerified: false,
      attestation: attestationResult,
    };
  }

  const requireProofVerification = options.requireProofVerification ?? true;
  let proofVerified = options.proofVerifiedHint === true;

  if (!proofVerified && options.proofVerifier) {
    try {
      proofVerified =
        (await options.proofVerifier({
          proof: envelope.proof,
          publicSignals: envelope.publicSignals,
          publicSignalsHash: envelope.publicSignalsHash,
          programHash: envelope.programHash,
          vkHash: envelope.vkHash,
        })) === true;
    } catch (error: unknown) {
      return {
        valid: false,
        reason: `Proof verifier threw an error: ${String(error)}`,
        proofVerified: false,
        attestation: attestationResult,
      };
    }
  }

  if (!proofVerified && requireProofVerification) {
    return {
      valid: false,
      reason:
        'ZK proof verification was required but no verifier/hint marked the proof as valid.',
      proofVerified: false,
      attestation: attestationResult,
    };
  }

  return {
    valid: true,
    proofVerified,
    attestation: attestationResult,
  };
}
