// @a0n/.ignored_gnosis/forkjoin-app/client/components/Hero3D/HeroScene.tsx
import { useRef, useMemo, useState, useEffect } from '@a0n/raect';
import { useFrame } from '@a0n/aeon-3d/fiber';
import { Float, MeshDistortMaterial, Stars, Trail } from '@a0n/aeon-3d/drei';
import * as THREE from '@a0n/aeon-3d/three';
import type { Visitor, CollectiveState } from '../../hooks/usePresence';

interface HeroSceneProps {
  visitors: Visitor[];
  collective: CollectiveState;
  localCursor: { x: number; y: number };
}

const MAX_RENDERED_VISITOR_LAYERS = 3;
const VISITOR_PARTICLE_CAP = 48;

/**
 * Renders the Hero Scene view.
 */
export function HeroScene({
  visitors,
  collective,
  localCursor,
}: HeroSceneProps) {
  // Track shooting stars for new visitor joins
  const [shootingStars, setShootingStars] = useState<
    Array<{ id: string; color: string; startTime: number }>
  >([]);
  const prevVisitorCount = useRef(visitors.length);
  const visibleVisitors = visitors.slice(0, MAX_RENDERED_VISITOR_LAYERS);
  const totalVisitorLayers = Math.min(
    visibleVisitors.length + 1,
    MAX_RENDERED_VISITOR_LAYERS + 1
  );

  // Detect new visitors and spawn shooting stars
  useEffect(() => {
    if (visitors.length > prevVisitorCount.current) {
      // New visitor(s) joined - spawn shooting star for each
      const newVisitors = visitors.slice(prevVisitorCount.current);
      const newStars = newVisitors.map((v) => ({
        id: `star-${v.id}-${Date.now()}`,
        color: v.color,
        startTime: Date.now(),
      }));
      setShootingStars((prev) => [...prev, ...newStars]);
    }
    prevVisitorCount.current = visitors.length;
  }, [visitors]);

  // Clean up old shooting stars after animation completes
  useEffect(() => {
    const cleanup = setInterval(() => {
      const now = Date.now();
      setShootingStars((prev) =>
        prev.filter((star) => now - star.startTime < 2000)
      );
    }, 1000);
    return () => clearInterval(cleanup);
  }, []);

  // Calculate activity from visitors' cursor movements
  const visitorActivity = useMemo(() => {
    let activity = 0;
    // Add local cursor activity
    activity += Math.abs(localCursor.x - 0.5) + Math.abs(localCursor.y - 0.5);

    visitors.forEach((v) => {
      if (v.cursor) {
        // More activity = more slosh
        activity += Math.abs(v.cursor.x - 0.5) + Math.abs(v.cursor.y - 0.5);
      }
    });
    return Math.min(activity / Math.max(visitors.length + 1, 1), 1);
  }, [visitors, localCursor]);

  return (
    <>
      {/* Fog to fade stars to black at edges */}
      <fog attach="fog" args={['#0a0a0f', 30, 100]} />

      {/* Point lights */}
      <pointLight position={[10, 10, 10]} intensity={0.6} />
      <pointLight position={[-5, -5, 5]} intensity={0.3} color="#ff6b6b" />
      <pointLight position={[5, -5, -5]} intensity={0.3} color="#4ecdc4" />

      {/* Background stars - layered for twinkling effect */}
      <Stars
        radius={80}
        depth={40}
        count={700}
        factor={4}
        saturation={0}
        fade
        speed={0.5}
      />
      {/* Second layer with different speed for twinkling depth */}
      <Stars
        radius={60}
        depth={30}
        count={400}
        factor={2}
        saturation={0.1}
        fade
        speed={0.65}
      />
      {/* Subtle twinkling ambient light */}
      <TwinkleLight />

      {/* Central collective orb - main sphere */}
      <CollectivePulse
        collective={collective}
        visitorCount={visitors.length}
        activity={visitorActivity}
      />

      {/* Rainbow sloshing layers - one per visitor up to 5 */}
      <VisitorSloshLayer
        visitor={{
          id: 'local',
          color: '#ffffff',
          cursor: localCursor,
          region: 'local',
        }}
        index={0}
        totalVisitors={totalVisitorLayers}
      />
      {visibleVisitors.map((visitor, index) => (
        <VisitorSloshLayer
          key={visitor.id}
          visitor={visitor}
          index={index + 1}
          totalVisitors={totalVisitorLayers}
        />
      ))}

      {/* Visitor particles */}
      <VisitorParticles visitors={visitors} />

      {/* Shooting stars for new visitor joins */}
      {shootingStars.map((star) => (
        <ShootingStar key={star.id} color={star.color} />
      ))}
    </>
  );
}

/**
 * ShootingStar - Animated streak across the sky when a visitor joins
 */
function ShootingStar({ color }: { color: string }) {
  const meshRef = useRef<THREE.Mesh>(null);
  const startPos = useRef<THREE.Vector3>(null);
  const direction = useRef<THREE.Vector3>(null);
  const startTime = useRef(Date.now());

  // Random start position and direction
  useMemo(() => {
    // Start from random edge of the scene
    const angle = Math.random() * Math.PI * 2;
    const startRadius = 15 + Math.random() * 10;
    startPos.current = new THREE.Vector3(
      Math.cos(angle) * startRadius,
      5 + Math.random() * 10,
      Math.sin(angle) * startRadius * 0.3 - 5
    );

    // Direction toward center with some randomness
    direction.current = new THREE.Vector3(
      -startPos.current.x * 0.8 + (Math.random() - 0.5) * 5,
      -startPos.current.y * 0.6 - 3,
      -startPos.current.z * 0.5
    ).normalize();
  }, []);

  useFrame(() => {
    if (!meshRef.current || !startPos.current || !direction.current) return;

    const elapsed = (Date.now() - startTime.current) / 1000;
    const speed = 25;
    const t = elapsed * speed;

    // Move along the direction
    meshRef.current.position
      .copy(startPos.current)
      .addScaledVector(direction.current, t);

    // Fade out over time
    const material = meshRef.current.material as THREE.MeshBasicMaterial;
    const fadeProgress = Math.min(elapsed / 1.5, 1);
    material.opacity = 1 - fadeProgress;

    // Scale down as it fades
    const scale = 0.15 * (1 - fadeProgress * 0.5);
    meshRef.current.scale.setScalar(scale);
  });

  return (
    <Trail width={2} length={8} color={color} attenuation={(t: number) => t * t}>
      <mesh ref={meshRef}>
        <sphereGeometry args={[1, 8, 8]} />
        <meshBasicMaterial color={color} transparent opacity={1} />
      </mesh>
    </Trail>
  );
}

/**
 * TwinkleLight - Subtle pulsing ambient light for star twinkling effect
 */
function TwinkleLight() {
  const lightRef = useRef<THREE.AmbientLight>(null);

  useFrame((state) => {
    if (!lightRef.current) return;
    const time = state.clock.elapsedTime;

    // Multiple overlapping sine waves for organic twinkling
    const twinkle =
      Math.sin(time * 0.7) * 0.02 +
      Math.sin(time * 1.3) * 0.015 +
      Math.sin(time * 2.1) * 0.01;

    lightRef.current.intensity = 0.3 + twinkle;
  });

  return <ambientLight ref={lightRef} intensity={0.3} />;
}

interface CollectivePulseProps {
  collective: CollectiveState;
  visitorCount: number;
  activity: number;
}

function CollectivePulse({
  collective,
  visitorCount,
  activity,
}: CollectivePulseProps) {
  const meshRef = useRef<THREE.Mesh>(null);
  const materialRef = useRef<THREE.MeshPhysicalMaterial | null>(null);

  // Smoothed activity for easing
  const smoothedActivity = useRef(0);

  // Rainbow color cycling - faster with more visitors
  useFrame((state, delta) => {
    if (!meshRef.current) return;
    const time = state.clock.elapsedTime;

    // Lerp factor - lower = smoother/slower easing
    const lerpFactor = 1 - Math.pow(0.03, delta);

    // Smooth the activity value
    smoothedActivity.current +=
      (activity - smoothedActivity.current) * lerpFactor;
    const smoothActivity = smoothedActivity.current;

    // Rainbow cycle speed increases with visitors
    const cycleSpeed = 0.08 + visitorCount * 0.02;
    const hueShift = (time * cycleSpeed) % 1;

    // More dramatic breathing with more visitors
    const breatheIntensity = 0.1 + visitorCount * 0.02 + smoothActivity * 0.2;
    const breathe = Math.sin(time * 0.5) * breatheIntensity;

    // Scale based on activity AND visitor count
    const baseScale =
      0.7 + visitorCount * 0.05 + collective.interactionIntensity * 0.3;
    meshRef.current.scale.setScalar(baseScale + breathe);

    // More sloshy rotation with smoothed activity
    const rotationSpeed = 0.1 + smoothActivity * 0.3;
    meshRef.current.rotation.y = time * rotationSpeed;
    meshRef.current.rotation.x =
      Math.sin(time * 0.2 + smoothActivity) * (0.1 + smoothActivity * 0.2);
    meshRef.current.rotation.z = Math.cos(time * 0.15) * smoothActivity * 0.15;

    const saturation = 0.6 + collective.arousal * 0.3;
    const lightness = 0.45 + collective.interactionIntensity * 0.15;
    const dynamicHue = (0.58 + collective.valence * 0.05 + hueShift) % 1;
    if (materialRef.current) {
      materialRef.current.color.setHSL(dynamicHue, saturation, lightness);
      materialRef.current.emissive.setHSL(
        dynamicHue,
        saturation,
        Math.min(lightness + 0.08, 0.85)
      );
      materialRef.current.emissiveIntensity = 0.3 + smoothActivity * 0.25;
    }
  });

  // Distortion increases dramatically with visitors and activity
  const distortAmount =
    0.2 + visitorCount * 0.05 + activity * 0.4 + collective.arousal * 0.2;

  return (
    <Float
      speed={1 + activity}
      rotationIntensity={0.2 + activity * 0.3}
      floatIntensity={0.5 + activity * 0.3}
    >
      <mesh ref={meshRef}>
        <sphereGeometry args={[1, 40, 40]} />
        <MeshDistortMaterial
          ref={(value: THREE.MeshPhysicalMaterial | null) => {
            materialRef.current = value;
          }}
          color="#5f9ccf"
          emissive="#5f9ccf"
          emissiveIntensity={0.3}
          distort={Math.min(distortAmount, 0.6)}
          speed={2 + visitorCount * 0.5 + activity * 3}
          roughness={0.2}
          metalness={0.8}
        />
      </mesh>
    </Float>
  );
}

/**
 * VisitorSloshLayer - Rainbow sloshing sphere layer per visitor
 * Each visitor adds a unique colored layer that sloshes independently
 */
interface VisitorSloshLayerProps {
  visitor: Visitor;
  index: number;
  totalVisitors: number;
}

function VisitorSloshLayer({
  visitor,
  index,
  totalVisitors,
}: VisitorSloshLayerProps) {
  const meshRef = useRef<THREE.Mesh>(null);

  // Smoothed cursor values for easing
  const smoothedCursor = useRef({ x: 0, y: 0 });
  const smoothedActivity = useRef(0);

  // Each visitor gets a unique hue in rainbow spectrum
  const visitorHue = useMemo(() => {
    // Parse visitor color or use index-based hue
    const color = new THREE.Color(visitor.color);
    const hsl = { h: 0, s: 0, l: 0 };
    color.getHSL(hsl);
    return hsl.h;
  }, [visitor.color]);

  useFrame((state, delta) => {
    if (!meshRef.current) return;
    const time = state.clock.elapsedTime;

    // Lerp factor - lower = smoother/slower easing
    const lerpFactor = 1 - Math.pow(0.05, delta);

    // Each layer has unique phase and speed
    const phase = (index / totalVisitors) * Math.PI * 2;
    const speed = 0.3 + index * 0.1;

    // Target cursor values
    const targetX = visitor.cursor ? (visitor.cursor.x - 0.5) * 0.3 : 0;
    const targetY = visitor.cursor ? (visitor.cursor.y - 0.5) * 0.3 : 0;

    // Smoothly interpolate toward target
    smoothedCursor.current.x +=
      (targetX - smoothedCursor.current.x) * lerpFactor;
    smoothedCursor.current.y +=
      (targetY - smoothedCursor.current.y) * lerpFactor;

    // Layer offset - each layer orbits slightly differently
    const orbitRadius = 0.1 + index * 0.05;
    meshRef.current.position.set(
      Math.sin(time * speed + phase) * orbitRadius + smoothedCursor.current.x,
      Math.cos(time * speed * 0.7 + phase) * orbitRadius +
        smoothedCursor.current.y,
      Math.sin(time * speed * 0.5 + phase) * orbitRadius * 0.5
    );

    // Smooth cursor activity
    const targetActivity = visitor.cursor
      ? Math.abs(visitor.cursor.x - 0.5) + Math.abs(visitor.cursor.y - 0.5)
      : 0;
    smoothedActivity.current +=
      (targetActivity - smoothedActivity.current) * lerpFactor;

    const baseScale = 1.1 + index * 0.15;
    const breathe =
      Math.sin(time * 0.8 + phase) * (0.05 + smoothedActivity.current * 0.1);
    meshRef.current.scale.setScalar(baseScale + breathe);

    // Rotation influenced by smoothed cursor
    meshRef.current.rotation.y = time * 0.05 + smoothedCursor.current.x * 2;
    meshRef.current.rotation.x =
      Math.sin(time * 0.1 + phase) * 0.1 + smoothedCursor.current.y * 2;
  });

  // Rainbow color for this visitor
  const color = useMemo(() => {
    return new THREE.Color().setHSL(visitorHue, 0.7, 0.5);
  }, [visitorHue]);

  return (
    <mesh ref={meshRef}>
      <sphereGeometry args={[1, 20, 20]} />
      <MeshDistortMaterial
        color={color}
        emissive={color}
        emissiveIntensity={0.2}
        distort={0.12 + index * 0.04}
        speed={1.2 + index * 0.25}
        roughness={0.3}
        metalness={0.6}
        transparent
        opacity={0.15 - index * 0.02}
      />
    </mesh>
  );
}

interface VisitorParticlesProps {
  visitors: Visitor[];
}

function VisitorParticles({ visitors }: VisitorParticlesProps) {
  const instancedRef = useRef<THREE.InstancedMesh>(null);
  const tempObject = useMemo(() => new THREE.Object3D(), []);
  const tempColor = useMemo(() => new THREE.Color(), []);

  // Smoothed cursor offsets per particle
  const smoothedOffsets = useRef<Array<{ x: number; y: number }>>(
    Array.from({ length: VISITOR_PARTICLE_CAP }, () => ({ x: 0, y: 0 }))
  );

  // Create positions on a sphere surface
  const positions = useMemo(() => {
    const positions: THREE.Vector3[] = [];
    for (let i = 0; i < VISITOR_PARTICLE_CAP; i++) {
      // Fibonacci sphere distribution
      const phi = Math.acos(-1 + (2 * i) / VISITOR_PARTICLE_CAP);
      const theta = Math.sqrt(VISITOR_PARTICLE_CAP * Math.PI) * phi;
      const radius = 2.5 + Math.random() * 0.5;

      positions.push(
        new THREE.Vector3(
          radius * Math.sin(phi) * Math.cos(theta),
          radius * Math.sin(phi) * Math.sin(theta),
          radius * Math.cos(phi)
        )
      );
    }
    return positions;
  }, []);

  useFrame((state, delta) => {
    if (!instancedRef.current) return;
    const time = state.clock.elapsedTime;

    // Lerp factor for smooth easing
    const lerpFactor = 1 - Math.pow(0.05, delta);

    // Update instance transforms and colors
    for (let i = 0; i < Math.min(visitors.length, VISITOR_PARTICLE_CAP); i++) {
      const visitor = visitors[i];
      const basePos = positions[i % positions.length];

      // Target cursor offset
      const targetX = visitor.cursor ? (visitor.cursor.x - 0.5) * 0.5 : 0;
      const targetY = visitor.cursor ? (visitor.cursor.y - 0.5) * 0.5 : 0;

      // Smooth interpolation
      smoothedOffsets.current[i].x +=
        (targetX - smoothedOffsets.current[i].x) * lerpFactor;
      smoothedOffsets.current[i].y +=
        (targetY - smoothedOffsets.current[i].y) * lerpFactor;

      // Animate position with subtle movement
      const wobble = Math.sin(time * 2 + i) * 0.05;
      tempObject.position.set(
        basePos.x + smoothedOffsets.current[i].x + wobble,
        basePos.y + smoothedOffsets.current[i].y,
        basePos.z + Math.cos(time + i) * 0.05
      );

      // Scale pulsing
      const scale = 0.03 + Math.sin(time * 3 + i) * 0.01;
      tempObject.scale.setScalar(scale);

      tempObject.updateMatrix();
      instancedRef.current.setMatrixAt(i, tempObject.matrix);

      // Set color from visitor
      tempColor.setStyle(visitor.color);
      instancedRef.current.setColorAt(i, tempColor);
    }

    instancedRef.current.instanceMatrix.needsUpdate = true;
    if (instancedRef.current.instanceColor) {
      instancedRef.current.instanceColor.needsUpdate = true;
    }
  });

  return (
    <instancedMesh
      ref={instancedRef}
      args={[undefined, undefined, VISITOR_PARTICLE_CAP]}
    >
      <sphereGeometry args={[1, 10, 10]} />
      <meshStandardMaterial
        emissive="white"
        emissiveIntensity={0.5}
        toneMapped={false}
      />
    </instancedMesh>
  );
}
