// @a0n/.ignored_gnosis/forkjoin-app/client/components/Hero3D/LazyHeroCanvas.tsx
/**
 * LazyHeroCanvas - Dynamically imported wrapper for the Three.js hero scene.
 * Keeps @a0n/aeon-3d/fiber and three out of the initial bundle.
 */
import { Suspense } from '@a0n/raect';
import { Canvas } from '@a0n/aeon-3d/fiber';
import { HeroScene } from './HeroScene';
import type { Visitor, CollectiveState } from '../../hooks/usePresence';

interface LazyHeroCanvasProps {
  visitors: Visitor[];
  collective: CollectiveState;
  localCursor: { x: number; y: number };
}

/**
 * Renders the Lazy Hero Canvas view.
 */
export default function LazyHeroCanvas({
  visitors,
  collective,
  localCursor,
}: LazyHeroCanvasProps) {
  return (
    <Canvas
      camera={{ position: [0, 0, 5], fov: 60 }}
      dpr={[1, 1.5]}
      gl={{ antialias: false, powerPreference: 'high-performance' }}
      style={{
        position: 'absolute',
        inset: 0,
        background: 'linear-gradient(135deg, #0a0a0f 0%, #12121a 100%)',
      }}
    >
      <Suspense fallback={null}>
        <HeroScene
          visitors={visitors}
          collective={collective}
          localCursor={localCursor}
        />
      </Suspense>
    </Canvas>
  );
}
