// @a0n/.ignored_gnosis/src/betty/quantum/bridge.ts
declare const require: ((specifier: string) => unknown) | undefined;

function tryRequire(specifier: string): unknown {
  if (typeof require !== 'function') return undefined;
  try {
    return require(specifier);
  } catch {
    return undefined;
  }
}

const GnosisRuntime =
  tryRequire('gnosis_runtime') ??
  tryRequire('../../../node_modules/gnosis_runtime');

interface QuantumRuntimeInstance {
  process_frame?: (encodedBytes: Uint8Array) => Uint8Array | void;
  processFrame?: (encodedBytes: Uint8Array) => Uint8Array | void;
  metrics: () => string;
}

interface GnosisRuntimeModule {
  QuantumRuntime?: new () => QuantumRuntimeInstance;
}

export class QuantumWasmBridge {
  private runtime: QuantumRuntimeInstance | null = null;
  private initialized = false;

  constructor() {
    try {
      const runtimeModule = GnosisRuntime as GnosisRuntimeModule | undefined;
      if (runtimeModule?.QuantumRuntime) {
        this.runtime = new runtimeModule.QuantumRuntime();
        this.initialized =
          typeof this.runtime.process_frame === 'function' ||
          typeof this.runtime.processFrame === 'function';
      }
    } catch {
      // Ignored
    }
  }

  public processAstEdge(
    edgeType: string,
    sourceCount: number,
    targetCount: number
  ): string {
    if (!this.initialized)
      return `[WASM] Not initialized or process_frame missing.`;

    let flags = 0;
    switch (edgeType) {
      case 'FORK':
        flags = 0b00000001;
        break;
      case 'RACE':
        flags = 0b00000010;
        break;
      case 'FOLD':
      case 'COLLAPSE':
        flags = 0b00000100;
        break;
      case 'VENT':
      case 'TUNNEL':
        flags = 0b00001000;
        break;
      case 'SLIVER':
        flags = 0b00100000;
        break;
      case 'OBSERVE':
        flags = 0b01000000;
        break;
      default:
        return `[WASM] Ignored edge type: ${edgeType}`;
    }

    const payload = new TextEncoder().encode(
      JSON.stringify({ s: sourceCount, t: targetCount })
    );
    const len = payload.length;

    // Manual 10-byte Aeon Flow Frame Encoding (Big Endian)
    const buffer = new Uint8Array(10 + len);
    const view = new DataView(buffer.buffer);

    view.setUint16(0, 1, false); // Stream ID: 1
    view.setUint32(2, 1, false); // Sequence: 1
    view.setUint8(6, flags); // Flags

    // 24-bit length (big endian)
    buffer[7] = (len >> 16) & 0xff;
    buffer[8] = (len >> 8) & 0xff;
    buffer[9] = len & 0xff;

    buffer.set(payload, 10);

    try {
      if (this.runtime && typeof this.runtime.process_frame === 'function') {
        this.runtime.process_frame(buffer);
      } else {
        this.runtime?.processFrame?.(buffer);
      }
      return `[WASM] Processed ${edgeType}. Metrics: ${
        this.runtime?.metrics() ?? 'unavailable'
      }`;
    } catch (error: unknown) {
      return `[WASM Error] ${this.errorMessage(error)}`;
    }
  }

  public getMetrics(): string {
    return this.initialized && this.runtime
      ? this.runtime.metrics()
      : 'Not initialized';
  }

  private errorMessage(error: unknown): string {
    if (error instanceof Error) {
      return error.message;
    }
    return String(error);
  }
}
