// @a0n/gnosis/forkjoin-app/src/mcp/manifests.ts
import {
  buildPublicSkillsIndex,
  type PublicSkillsIndex,
  type SkillDiscoveryEntry,
} from '@a0n/shared-utils/services/skills';
import { GNOSIS_MCP_TOOLS } from './tools';

const MCP_PROTOCOL_VERSION = '2024-11-05';
const SERVER_NAME = 'forkjoin-app';
const SERVER_VERSION = '0.1.0';

export const GNOSIS_FEATURES = [
  {
    id: 'gnosis-analysis',
    name: 'Compile and Verify',
    description:
      'Compile, lint, analyze, verify, generate TLA, and emit SARIF for Gnosis source.',
    tools: [
      'gnosis_compile',
      'gnosis_lint',
      'gnosis_analyze',
      'gnosis_verify',
      'gnosis_generate_tla',
      'gnosis_emit_sarif',
    ],
    resources: ['gnosis://tools/public'],
  },
  {
    id: 'gnosis-collaboration',
    name: 'QDoc + DashRelay Collaboration',
    description:
      'Collaborative CRDT sessions via QDoc operations and DashRelay session metadata.',
    tools: [
      'qdoc_create',
      'qdoc_get_state',
      'qdoc_apply_update',
      'qdoc_transact',
      'qdoc_get_delta',
      'dashrelay_connect_qdoc',
      'dashrelay_status',
      'dashrelay_disconnect',
    ],
    resources: ['gnosis://dashrelay'],
  },
  {
    id: 'gnosis-runtime-and-deploy',
    name: 'Runtime and Deploy Hooks',
    description:
      'Run interpreted/native Gnosis and optionally trigger Aeon Forge deploy hooks.',
    tools: [
      'gnosis_run',
      'gnosis_native_run',
      'gnosis_test_file',
      'gnosis_test_suite',
      'gnosis_forge_deploy',
    ],
    resources: ['gnosis://meta'],
  },
  {
    id: 'buleyean-logic-and-proof',
    name: 'Buleyean Logic, Distribution & Proof',
    description:
      'Post-fold logic where truth is ground state. Compute complement distributions from void boundaries, apply connectives (AND/OR/NOT/IMPLIES/XOR/FORALL/EXISTS), run proof-by-rejection, verify proof topologies, and emit Lean 4 or TLA+ artifacts. Boolean is the K=2 special case.',
    tools: [
      'buleyean_distribution',
      'buleyean_axioms',
      'buleyean_deficit',
      'buleyean_proposition',
      'buleyean_connective',
      'buleyean_proof',
      'buleyean_boolean_embed',
      'buleyean_verify_proof_topology',
      'buleyean_child_decides',
    ],
    resources: ['gnosis://buleyean'],
  },
  {
    id: 'skyrms-restoration',
    name: 'Skyrms Restoration Triage',
    description:
      'O(1) checkpoint restoration triage via God Formula Skyrms equilibrium. Sovereign (truth-establishing, v=0) before oracle (void-mapping, v>=R). Each checkpoint scored by restorationValue = beta1 * godWeight(R, v). For cancer genomics, HFT helix repair, or any system with degraded checkpoint topology. For Sandy.',
    tools: ['skyrms_restoration_triage'],
    resources: [],
  },
] as const;

export const GNOSIS_EXAMPLES = [
  {
    id: 'skyrms-triage-gbm',
    prompt:
      'Triage checkpoint restoration for GBM Combined (p53, Rb, APC knocked out). Return the three-phase Skyrms restoration sequence.',
    outcome:
      'Phase 1 sovereign: p53 (godWeight=10, value=30). Phase 2 midweight: Rb (godWeight=7, value=14). Phase 3 midweight: APC (godWeight=3, value=6). Total cost: 3 clinamen.',
  },
  {
    id: 'compile-to-sarif',
    prompt:
      'Compile Gnosis source, verify it against workers target, and return SARIF plus TLA output.',
    outcome: 'Static analysis + formal verification artifacts in one pass.',
  },
  {
    id: 'collaborative-qdoc-flow',
    prompt:
      'Create a QDoc session, apply map/text operations, inspect state, and publish relay room metadata.',
    outcome: 'CRDT collaboration state with relay session continuity.',
  },
  {
    id: 'forge-deploy-gate',
    prompt:
      'Run gnosis_test_suite then trigger gnosis_forge_deploy only when all checks pass.',
    outcome: 'Deployment hooks gated by verification status.',
  },
  {
    id: 'buleyean-proof-by-rejection',
    prompt:
      'Create a proof goal with 3 propositions, apply rejection steps until all reach ground state, then render the proof trace and verify the three axioms hold.',
    outcome: 'Complete Buleyean proof with QED status and axiom verification.',
  },
  {
    id: 'buleyean-proof-topology',
    prompt:
      'Write a .gg proof topology with Theorem/Axiom nodes and REJECT edges, verify it with buleyean_verify_proof_topology, and emit Lean 4 artifacts.',
    outcome: 'Verified proof topology with mechanized Lean 4 proof artifact.',
  },
] as const;

export const GNOSIS_SKILLS = [
  {
    id: 'gnosis-verification-workbench',
    name: 'Gnosis Verification Workbench',
    uri: 'skill://forkjoin-app/verification-workbench',
    description:
      'Compile, lint, verify, generate TLA, and emit SARIF in a single verification workflow.',
    tools: [
      'gnosis_compile',
      'gnosis_lint',
      'gnosis_verify',
      'gnosis_generate_tla',
      'gnosis_emit_sarif',
    ],
    tags: ['gnosis', 'verification', 'tla', 'sarif'],
    source: 'local',
  },
  {
    id: 'gnosis-crdt-collab',
    name: 'Gnosis CRDT Collaboration',
    uri: 'skill://forkjoin-app/crdt-collab',
    description:
      'Create and update QDoc collaborative state, then bridge metadata to DashRelay sessions.',
    tools: [
      'qdoc_create',
      'qdoc_transact',
      'qdoc_get_state',
      'qdoc_get_delta',
      'dashrelay_connect_qdoc',
      'dashrelay_status',
    ],
    tags: ['gnosis', 'crdt', 'qdoc', 'dashrelay'],
    source: 'local',
  },
  {
    id: 'gnosis-runtime-to-forge',
    name: 'Runtime to Forge Deploy',
    uri: 'skill://forkjoin-app/runtime-to-forge',
    description:
      'Execute/test gnosis programs and trigger Aeon Forge deploy hooks when authorized.',
    tools: [
      'gnosis_run',
      'gnosis_native_run',
      'gnosis_test_file',
      'gnosis_test_suite',
      'gnosis_forge_deploy',
    ],
    tags: ['gnosis', 'runtime', 'deploy', 'forge'],
    source: 'local',
  },
  {
    id: 'buleyean-proof-workbench',
    name: 'Buleyean Proof Workbench',
    uri: 'skill://forkjoin-app/buleyean-proof-workbench',
    description:
      'Compute Buleyean distributions, apply logic connectives, run proofs by rejection, verify proof topologies, and emit formal artifacts (Lean 4, TLA+).',
    tools: [
      'buleyean_distribution',
      'buleyean_axioms',
      'buleyean_proposition',
      'buleyean_connective',
      'buleyean_proof',
      'buleyean_verify_proof_topology',
    ],
    tags: ['buleyean', 'logic', 'proof', 'lean4', 'tla', 'void-boundary'],
    source: 'local',
  },
] as const satisfies ReadonlyArray<SkillDiscoveryEntry>;

export interface DiscoveryEndpoints {
  origin: string;
  mcp: string;
  llms: string;
  llmsFull: string;
  skills: string;
  skillsIndex: string;
  agents: string;
  aiPlugin: string;
  mcpWellKnown: string;
  mcpManifest: string;
  agentCard: string;
  openApi: string;
  security: string;
}

function normalizeBaseUrl(value: string): string {
  return value.replace(/\/+$/, '');
}

/**
 * Resolves the Public Base Url.
 */
export function resolvePublicBaseUrl(
  request: Request,
  configured?: string
): string {
  if (typeof configured === 'string' && configured.trim().length > 0) {
    return normalizeBaseUrl(configured.trim());
  }

  const url = new URL(request.url);
  return url.origin;
}

/**
 * Builds the Discovery Endpoints.
 */
export function buildDiscoveryEndpoints(
  request: Request,
  configuredBaseUrl?: string
): DiscoveryEndpoints {
  const origin = resolvePublicBaseUrl(request, configuredBaseUrl);
  return {
    origin,
    mcp: `${origin}/mcp`,
    llms: `${origin}/llms.txt`,
    llmsFull: `${origin}/llms-full.txt`,
    skills: `${origin}/skills`,
    skillsIndex: `${origin}/skills/index.json`,
    agents: `${origin}/agents.json`,
    aiPlugin: `${origin}/.well-known/ai-plugin.json`,
    mcpWellKnown: `${origin}/.well-known/mcp.json`,
    mcpManifest: `${origin}/.well-known/mcp-manifest.json`,
    agentCard: `${origin}/.well-known/agent-card.json`,
    openApi: `${origin}/.well-known/openapi.json`,
    security: `${origin}/.well-known/security.txt`,
  };
}

/**
 * Builds the Llms Txt.
 */
export function buildLlmsTxt(endpoints: DiscoveryEndpoints): string {
  const publicTools = GNOSIS_MCP_TOOLS.filter(
    (tool) => tool.access === 'public'
  );
  const authTools = GNOSIS_MCP_TOOLS.filter((tool) => tool.access === 'auth');

  return [
    '# ForkJoin Unified Programming + MCP Platform',
    '',
    'Model Context Protocol endpoint for compile/lint/verify tooling, CRDT collaboration, and deploy hooks.',
    '',
    `MCP endpoint: ${endpoints.mcp}`,
    `Skills index: ${endpoints.skillsIndex}`,
    `Agent manifest: ${endpoints.agents}`,
    '',
    `Public tools (${publicTools.length}):`,
    ...publicTools.map((tool) => `- ${tool.name}`),
    '',
    `Auth tools (${authTools.length}):`,
    ...authTools.map((tool) => `- ${tool.name}`),
    '',
    'Web routes:',
    '- /',
    '- /ide',
    '- /repl',
    '- /playground',
    '- /docs',
  ].join('\n');
}

/**
 * Builds the Llms Full Txt.
 */
export function buildLlmsFullTxt(endpoints: DiscoveryEndpoints): string {
  return [
    '# ForkJoin MCP Full Metadata',
    '',
    `Name: ${SERVER_NAME}`,
    `Version: ${SERVER_VERSION}`,
    `Protocol: ${MCP_PROTOCOL_VERSION}`,
    '',
    `MCP: ${endpoints.mcp}`,
    `LLMS: ${endpoints.llms}`,
    `LLMS Full: ${endpoints.llmsFull}`,
    `Skills: ${endpoints.skills}`,
    `Skills Index: ${endpoints.skillsIndex}`,
    `Agents: ${endpoints.agents}`,
    `AI Plugin: ${endpoints.aiPlugin}`,
    `MCP Manifest: ${endpoints.mcpWellKnown}`,
    `MCP Marketing: ${endpoints.mcpManifest}`,
    `Agent Card: ${endpoints.agentCard}`,
    `OpenAPI: ${endpoints.openApi}`,
    `Security: ${endpoints.security}`,
    '',
    'Features:',
    ...GNOSIS_FEATURES.map(
      (feature) => `- ${feature.name}: ${feature.description}`
    ),
    '',
    'Examples:',
    ...GNOSIS_EXAMPLES.map((example) => `- ${example.prompt}`),
  ].join('\n');
}

/**
 * Builds the Mcp Well Known Manifest.
 */
export function buildMcpWellKnownManifest(
  endpoints: DiscoveryEndpoints
): Record<string, unknown> {
  return {
    id: SERVER_NAME,
    name: 'ForkJoin MCP',
    version: SERVER_VERSION,
    protocolVersion: MCP_PROTOCOL_VERSION,
    endpoint: endpoints.mcp,
    llms: endpoints.llms,
    skills: endpoints.skillsIndex,
    agents: endpoints.agents,
    toolCount: GNOSIS_MCP_TOOLS.length,
    capabilities: {
      tools: true,
      resources: true,
    },
    features: GNOSIS_FEATURES,
  };
}

/**
 * Builds the Mcp Marketing Manifest.
 */
export function buildMcpMarketingManifest(
  endpoints: DiscoveryEndpoints
): Record<string, unknown> {
  return {
    schema_version: '1.0',
    server: {
      name: SERVER_NAME,
      version: SERVER_VERSION,
    },
    endpoint: endpoints.mcp,
    protocol_version: MCP_PROTOCOL_VERSION,
    transport: 'http',
    discovery: {
      llms: endpoints.llms,
      llmsFull: endpoints.llmsFull,
      skills: endpoints.skillsIndex,
      agents: endpoints.agents,
      manifest: endpoints.mcpWellKnown,
      agentCard: endpoints.agentCard,
      openApi: endpoints.openApi,
      aiPlugin: endpoints.aiPlugin,
      security: endpoints.security,
    },
    features: GNOSIS_FEATURES,
    examples: GNOSIS_EXAMPLES,
  };
}

/**
 * Builds the Agent Card Manifest.
 */
export function buildAgentCardManifest(
  endpoints: DiscoveryEndpoints
): Record<string, unknown> {
  return {
    id: 'forkjoin-app-agent-card',
    name: 'ForkJoin',
    description:
      'Unified programming site + MCP surface for Gnosis language, CRDT collaboration, and deploy hooks.',
    url: endpoints.origin,
    protocol: 'mcp',
    endpoint: endpoints.mcp,
    capabilities: {
      tools: GNOSIS_MCP_TOOLS.map((tool) => tool.name),
      protocols: ['mcp'],
      integrations: ['aeon-container-ide', 'aeon-forge', 'dashrelay'],
    },
    authentication: {
      schemes: ['bearer'],
      optionalForPublicTools: true,
    },
    featured_skills: GNOSIS_SKILLS.slice(0, 2).map((skill) => skill.id),
  };
}

/**
 * Builds the Agents Manifest.
 */
export function buildAgentsManifest(
  endpoints: DiscoveryEndpoints
): Record<string, unknown> {
  return {
    schema_version: '1.0.0',
    name: SERVER_NAME,
    description:
      'ForkJoin MCP server for compile/lint/verify/TLA/SARIF, CRDT collaboration, DashRelay sessioning, and deploy hooks.',
    endpoints: {
      mcp: endpoints.mcp,
      llms: endpoints.llms,
      llmsFull: endpoints.llmsFull,
      skills: endpoints.skillsIndex,
      manifest: endpoints.mcpWellKnown,
      agentCard: endpoints.agentCard,
      openApi: endpoints.openApi,
      security: endpoints.security,
    },
    capabilities: GNOSIS_FEATURES,
    examples: GNOSIS_EXAMPLES,
    onboarding: {
      initialize: {
        method: 'POST',
        endpoint: endpoints.mcp,
      },
      listToolsMethod: 'tools/list',
      callToolMethod: 'tools/call',
      listResourcesMethod: 'resources/list',
      readResourceMethod: 'resources/read',
    },
  };
}

/**
 * Builds the Ai Plugin Manifest.
 */
export function buildAiPluginManifest(
  endpoints: DiscoveryEndpoints
): Record<string, unknown> {
  return {
    schema_version: 'v1',
    name_for_human: 'ForkJoin MCP',
    name_for_model: SERVER_NAME,
    description_for_human:
      'Gnosis compile/verify + CRDT collaboration + deploy hooks over MCP.',
    description_for_model:
      'Model Context Protocol server with public analysis tools and auth-gated execution/collaboration/deploy tools.',
    auth: {
      type: 'service_http',
      authorization_type: 'bearer',
    },
    api: {
      type: 'openapi',
      url: endpoints.openApi,
      is_user_authenticated: false,
    },
    logo_url: `${endpoints.origin}/favicon.ico`,
    contact_email: 'support@forkjoin.ai',
    legal_info_url: `${endpoints.origin}/terms`,
  };
}

/**
 * Builds the Open Api Manifest.
 */
export function buildOpenApiManifest(
  endpoints: DiscoveryEndpoints
): Record<string, unknown> {
  return {
    openapi: '3.1.0',
    info: {
      title: 'ForkJoin MCP API',
      version: SERVER_VERSION,
      description:
        'Discovery + MCP JSON-RPC endpoint for the unified forkjoin-app worker.',
    },
    servers: [
      {
        url: endpoints.origin,
      },
    ],
    paths: {
      '/health': {
        get: {
          operationId: 'getHealth',
          summary: 'Get worker health details.',
          responses: {
            '200': {
              description: 'Service health payload.',
            },
          },
        },
      },
      '/mcp': {
        get: {
          operationId: 'getMcpInfo',
          summary: 'Get MCP endpoint summary.',
          responses: {
            '200': {
              description: 'MCP endpoint metadata.',
            },
          },
        },
        post: {
          operationId: 'postMcpJsonRpc',
          summary: 'Send MCP JSON-RPC request.',
          responses: {
            '200': {
              description: 'MCP JSON-RPC response.',
            },
          },
        },
      },
      '/llms.txt': {
        get: {
          operationId: 'getLlmsTxt',
          summary: 'Return llms onboarding text.',
          responses: {
            '200': {
              description: 'Plain text response.',
            },
          },
        },
      },
      '/llms-full.txt': {
        get: {
          operationId: 'getLlmsFullTxt',
          summary: 'Return full llms metadata text.',
          responses: {
            '200': {
              description: 'Plain text response.',
            },
          },
        },
      },
      '/agents.json': {
        get: {
          operationId: 'getAgentsManifest',
          summary: 'Return agent manifest.',
          responses: {
            '200': {
              description: 'Agent manifest JSON.',
            },
          },
        },
      },
      '/skills': {
        get: {
          operationId: 'getSkills',
          summary: 'Return MCP skill list payload.',
          responses: {
            '200': {
              description: 'Skills list JSON.',
            },
          },
        },
      },
      '/skills/index.json': {
        get: {
          operationId: 'getSkillsIndex',
          summary: 'Return public skills index manifest.',
          responses: {
            '200': {
              description: 'Skills index JSON.',
            },
          },
        },
      },
      '/.well-known/mcp.json': {
        get: {
          operationId: 'getMcpWellKnown',
          summary: 'Return machine-readable MCP manifest.',
          responses: {
            '200': {
              description: 'MCP manifest JSON.',
            },
          },
        },
      },
      '/.well-known/mcp-manifest.json': {
        get: {
          operationId: 'getMcpMarketingManifest',
          summary: 'Return marketing manifest for MCP listing.',
          responses: {
            '200': {
              description: 'Marketing manifest JSON.',
            },
          },
        },
      },
      '/.well-known/agent-card.json': {
        get: {
          operationId: 'getAgentCard',
          summary: 'Return agent card metadata.',
          responses: {
            '200': {
              description: 'Agent card JSON.',
            },
          },
        },
      },
      '/.well-known/openapi.json': {
        get: {
          operationId: 'getOpenApi',
          summary: 'Return OpenAPI manifest.',
          responses: {
            '200': {
              description: 'OpenAPI JSON.',
            },
          },
        },
      },
      '/.well-known/security.txt': {
        get: {
          operationId: 'getSecurityTxt',
          summary: 'Return security disclosure metadata.',
          responses: {
            '200': {
              description: 'Security text payload.',
            },
          },
        },
      },
    },
  };
}

/**
 * Builds the Skills Index.
 */
export function buildSkillsIndex(
  endpoints: DiscoveryEndpoints
): PublicSkillsIndex {
  return buildPublicSkillsIndex({
    server: {
      id: SERVER_NAME,
      name: 'ForkJoin MCP',
      version: SERVER_VERSION,
    },
    discovery: {
      endpoint: endpoints.mcp,
      llms: endpoints.llms,
      manifest: endpoints.mcpWellKnown,
      agentCard: endpoints.agentCard,
    },
    skills: GNOSIS_SKILLS,
    featuredSkills: GNOSIS_SKILLS.slice(0, 2).map((skill) => skill.id),
    features: GNOSIS_FEATURES,
    examples: GNOSIS_EXAMPLES,
  });
}

/**
 * Builds the Security Txt.
 */
export function buildSecurityTxt(endpoints: DiscoveryEndpoints): string {
  const expiresAt = new Date(
    Date.now() + 1000 * 60 * 60 * 24 * 365
  ).toISOString();

  return [
    'Contact: mailto:security@forkjoin.ai',
    `Canonical: ${endpoints.security}`,
    `Policy: ${endpoints.origin}/terms`,
    'Preferred-Languages: en',
    `Expires: ${expiresAt}`,
  ].join('\n');
}
