// @a0n/gnosis/forkjoin-app/src/mcp/resources.ts
import type { Env, McpResourceDefinition } from '../types';
import { AUTH_TOOL_NAMES, PUBLIC_TOOL_NAMES } from './tools';

export const GNOSIS_MCP_RESOURCES: McpResourceDefinition[] = [
  {
    uri: 'gnosis://meta',
    name: 'ForkJoin MCP Metadata',
    description: 'Server metadata, environment, and deployment defaults.',
    mimeType: 'application/json',
  },
  {
    uri: 'gnosis://docs-hub',
    name: 'Gnosis Documentation Hub',
    description:
      'Primary links for Gnosis, Aeon, Aeon Flux, Aeon Container, and Aeon Forge docs.',
    mimeType: 'application/json',
  },
  {
    uri: 'gnosis://legal',
    name: 'Legal and Policy Surface',
    description:
      'Privacy, cookies, cookie settings, terms, and colophon routes.',
    mimeType: 'application/json',
  },
  {
    uri: 'gnosis://dashrelay',
    name: 'DashRelay Defaults',
    description: 'DashRelay websocket, discovery, and webtransport defaults.',
    mimeType: 'application/json',
  },
  {
    uri: 'gnosis://tools/public',
    name: 'Public Tool Catalog',
    description: 'Public compile/lint/analyze/verify and linting tools.',
    mimeType: 'application/json',
  },
  {
    uri: 'gnosis://tools/auth',
    name: 'Auth-Gated Tool Catalog',
    description:
      'Execution, CRDT, relay, test, and deploy tools requiring auth.',
    mimeType: 'application/json',
  },
  {
    uri: 'gnosis://routes',
    name: 'HTTP Route Catalog',
    description: 'Web routes, MCP routes, and discovery endpoints.',
    mimeType: 'application/json',
  },
];

/**
 * Handles the gnosis read Gnosis Resource workflow.
 */
export function readGnosisResource(
  uri: string,
  env: Env,
  publicBaseUrl: string
): {
  contents: Array<{
    uri: string;
    mimeType: string;
    text: string;
  }>;
} {
  const normalizedBaseUrl = publicBaseUrl.replace(/\/+$/, '');

  if (uri === 'gnosis://meta') {
    return {
      contents: [
        {
          uri,
          mimeType: 'application/json',
          text: JSON.stringify(
            {
              service: 'forkjoin-app',
              protocol: 'mcp-json-rpc',
              environment: env.ENVIRONMENT,
              endpoint: `${normalizedBaseUrl}/mcp`,
              sessionHeader: 'mcp-session-id',
              hasAuthSecret:
                typeof env.MCP_AUTH_TOKEN === 'string' &&
                env.MCP_AUTH_TOKEN.trim().length > 0,
              publicToolRateLimitPerDay: env.PUBLIC_TOOL_RATE_LIMIT ?? '20',
            },
            null,
            2
          ),
        },
      ],
    };
  }

  if (uri === 'gnosis://docs-hub') {
    return {
      contents: [
        {
          uri,
          mimeType: 'application/json',
          text: JSON.stringify(
            {
              web: {
                home: `${normalizedBaseUrl}/`,
                ide: `${normalizedBaseUrl}/ide`,
                repl: `${normalizedBaseUrl}/repl`,
                playground: `${normalizedBaseUrl}/playground`,
                docs: `${normalizedBaseUrl}/docs`,
              },
              source: {
                gnosis: 'https://github.com/forkjoin-ai/gnosis',
                aeon: 'https://github.com/forkjoin-ai/aeon',
                aeonFlux: 'https://github.com/forkjoin-ai/aeon-flux',
                aeonContainer: 'https://github.com/forkjoin-ai/aeon-container',
                aeonForge: 'https://github.com/forkjoin-ai/aeon-forge',
              },
            },
            null,
            2
          ),
        },
      ],
    };
  }

  if (uri === 'gnosis://legal') {
    return {
      contents: [
        {
          uri,
          mimeType: 'application/json',
          text: JSON.stringify(
            {
              privacy: `${normalizedBaseUrl}/privacy`,
              cookies: `${normalizedBaseUrl}/cookies`,
              cookieSettings: `${normalizedBaseUrl}/cookie-settings`,
              terms: `${normalizedBaseUrl}/terms`,
              colophon: `${normalizedBaseUrl}/colophon`,
              copyright: 'All rights reserved.',
            },
            null,
            2
          ),
        },
      ],
    };
  }

  if (uri === 'gnosis://dashrelay') {
    return {
      contents: [
        {
          uri,
          mimeType: 'application/json',
          text: JSON.stringify(
            {
              relayUrl: env.DASH_RELAY_URL ?? 'https://relay.dashrelay.com',
              websocketUrl:
                env.DASH_RELAY_WS_URL ?? 'wss://relay.dashrelay.com/relay/sync',
              webtransportUrl:
                env.DASH_RELAY_WT_URL ?? 'https://relay.dashrelay.com/relay',
              discoveryUrl:
                env.DASH_RELAY_DISCOVERY_URL ??
                'https://relay.dashrelay.com/discovery',
              modes: ['hybrid', 'bring-your-own-room', 'ephemeral-fallback'],
            },
            null,
            2
          ),
        },
      ],
    };
  }

  if (uri === 'gnosis://tools/public') {
    return {
      contents: [
        {
          uri,
          mimeType: 'application/json',
          text: JSON.stringify(
            { tools: Array.from(PUBLIC_TOOL_NAMES).sort() },
            null,
            2
          ),
        },
      ],
    };
  }

  if (uri === 'gnosis://tools/auth') {
    return {
      contents: [
        {
          uri,
          mimeType: 'application/json',
          text: JSON.stringify(
            {
              tools: Array.from(AUTH_TOOL_NAMES).sort(),
              authRequired: true,
            },
            null,
            2
          ),
        },
      ],
    };
  }

  if (uri === 'gnosis://routes') {
    return {
      contents: [
        {
          uri,
          mimeType: 'application/json',
          text: JSON.stringify(
            {
              webRoutes: [
                '/',
                '/ide',
                '/repl',
                '/playground',
                '/docs',
                '/privacy',
                '/cookies',
                '/cookie-settings',
                '/colophon',
                '/terms',
              ],
              apiRoutes: ['/mcp', '/health'],
              discoveryRoutes: [
                '/llms.txt',
                '/llms-full.txt',
                '/skills',
                '/skills.json',
                '/skills/index.json',
                '/skills/:name.md',
                '/.well-known/ai-plugin.json',
                '/.well-known/mcp.json',
                '/.well-known/mcp-manifest.json',
                '/.well-known/agent-card.json',
                '/.well-known/openapi.json',
                '/.well-known/security.txt',
                '/agents.json',
              ],
            },
            null,
            2
          ),
        },
      ],
    };
  }

  throw new Error(`Unknown resource: ${uri}`);
}
