// @a0n/.ignored_gnosis/bin/bw.ts
#!/usr/bin/env node

/**
 * bw — Bitwise CLI tool for Lexicon binary compression and decompression.
 * 
 * Usage:
 *   bw compress <input> [output] [--type json|dna]
 *   bw decompress <input> [output]
 */

import fs from 'node:fs';
import path from 'node:path';
import crypto from 'node:crypto';
import { performance } from 'node:perf_hooks';
import { createTopologyCodec } from '../../dash/src/engine/topology-codec';

/**
 * Runs the gnosis command-line workflow.
 */
export async function main(args?: string[]) {
  if (!args) {
    args = process.argv.slice(2);
  }
  const cmd = args[0];

  if (!cmd || cmd === 'help' || cmd === '--help' || cmd === '-h') {
    printHelp();
    return;
  }

  try {
    if (['xxd', 'hexdump', 'od', 'strings', 'sha256sum', 'shasum', 'md5sum'].includes(cmd)) {
      await handleStandardBinary(cmd, args.slice(1));
    } else if (cmd === 'compress') {
      await handleCompress(args.slice(1));
    } else if (cmd === 'decompress') {
      await handleDecompress(args.slice(1));
    } else if (cmd === 'inspect') {
      await handleInspect(args.slice(1));
    } else {
      console.error(`Unknown command: ${cmd}`);
      printHelp();
      process.exit(1);
    }
  } catch (error) {
    console.error(`Error: ${error.message}`);
    process.exit(1);
  }
}

type Format = 'gg' | 'json' | 'plain';

interface BinaryTopology {
  command: string;
  path: string;
  bytes: number;
  startedAt: string;
  completedAt: string;
  readUs: number;
  transformUs: number;
  outputLines: string[];
  digest?: string;
  algorithm?: string;
}

function printHelp() {
  console.log(`
bw — Bitwise CLI tool for Lexicon binary compression

Usage:
  bw compress <input> [output] [--type json|dna]
  bw decompress <input> [output]
  bw inspect <input>
  bw xxd|hexdump|od|strings|sha256sum|shasum|md5sum <input> [--plain|--json|--gg]

Options:
  --type <type>    Type of data: json (default), dna (for 2-bit packing), or gg (to compile and shrink flowframes)
  --help, -h       Show this help message
`);
}

function parseFormat(args: string[]): { format: Format; args: string[] } {
  let format: Format = 'gg';
  const filtered: string[] = [];
  for (const arg of args) {
    if (arg === '--plain') format = 'plain';
    else if (arg === '--json') format = 'json';
    else if (arg === '--gg') format = 'gg';
    else filtered.push(arg);
  }
  return { format, args: filtered };
}

function esc(value: string): string {
  return value.replace(/\\/g, '\\\\').replace(/'/g, "\\'").replace(/\n/g, '\\n');
}

function hexRows(binary: Buffer, width = 16): string[] {
  const rows: string[] = [];
  for (let offset = 0; offset < binary.length; offset += width) {
    const chunk = binary.subarray(offset, offset + width);
    const hex = Array.from(chunk).map((byte) => byte.toString(16).padStart(2, '0')).join(' ');
    const ascii = Array.from(chunk).map((byte) => (byte >= 32 && byte <= 126 ? String.fromCharCode(byte) : '.')).join('');
    rows.push(`${offset.toString(16).padStart(8, '0')}: ${hex.padEnd(width * 3 - 1)}  ${ascii}`);
  }
  return rows;
}

function stringRows(binary: Buffer): string[] {
  return binary
    .toString('latin1')
    .split(/[^\x20-\x7e]+/g)
    .filter((value) => value.length >= 4);
}

function renderBinaryGG(topology: BinaryTopology): string {
  const lines = [
    `(bw: BinaryCommand { command: '${esc(topology.command)}', path: '${esc(topology.path)}', bytes: ${topology.bytes}, read_us: ${topology.readUs}, transform_us: ${topology.transformUs} })`,
    `(blob: ByteBlob { path: '${esc(topology.path)}', bytes: ${topology.bytes} })`,
    `(bw)-[:READ { elapsed_us: ${topology.readUs} }]->(blob)`,
  ];
  if (topology.digest && topology.algorithm) {
    lines.push(`(digest: Digest { algorithm: '${topology.algorithm}', value: '${topology.digest}' })`);
    lines.push(`(blob)-[:HASH { elapsed_us: ${topology.transformUs} }]->(digest)`);
    lines.push('(digest)-[:FOLD]->(bw)');
  }
  topology.outputLines.slice(0, 128).forEach((line, index) => {
    lines.push(`(row_${index}: BinaryRow { index: ${index}, text: '${esc(line)}' })`);
    lines.push(`(blob)-[:WINDOW { order: ${index} }]->(row_${index})`);
  });
  lines.push(`(stdout: Output { lines: ${topology.outputLines.length} })`);
  lines.push('(bw)-[:EMIT]->(stdout)');
  return `${lines.join('\n')}\n`;
}

function emitBinary(topology: BinaryTopology, format: Format) {
  if (format === 'json') {
    console.log(JSON.stringify(topology, null, 2));
  } else if (format === 'plain') {
    console.log(topology.outputLines.join('\n'));
  } else {
    process.stdout.write(renderBinaryGG(topology));
  }
}

async function handleStandardBinary(command: string, rawArgs: string[]) {
  const { format, args } = parseFormat(rawArgs);
  const input = args[0];
  if (!input) {
    throw new Error(`Missing input file for ${command}`);
  }

  const abs = path.resolve(input);
  const startedAt = new Date().toISOString();
  const readStarted = performance.now();
  const binary = fs.readFileSync(abs);
  const readUs = Math.round((performance.now() - readStarted) * 1000);
  const transformStarted = performance.now();
  let outputLines: string[] = [];
  let digest: string | undefined;
  let algorithm: string | undefined;

  if (command === 'xxd' || command === 'hexdump') {
    outputLines = hexRows(binary);
  } else if (command === 'od') {
    outputLines = Array.from(binary)
      .reduce<string[]>((rows, byte, index) => {
        if (index % 16 === 0) rows.push(index.toString(8).padStart(7, '0'));
        const last = rows.length - 1;
        rows[last] = `${rows[last] ?? ''} ${byte.toString(8).padStart(3, '0')}`;
        return rows;
      }, []);
  } else if (command === 'strings') {
    outputLines = stringRows(binary);
  } else {
    algorithm = command === 'md5sum' ? 'md5' : command === 'shasum' ? 'sha1' : 'sha256';
    digest = crypto.createHash(algorithm).update(binary).digest('hex');
    outputLines = [`${digest}  ${input}`];
  }

  const topology: BinaryTopology = {
    command,
    path: abs,
    bytes: binary.length,
    startedAt,
    completedAt: new Date().toISOString(),
    readUs,
    transformUs: Math.round((performance.now() - transformStarted) * 1000),
    outputLines,
  };
  if (digest) topology.digest = digest;
  if (algorithm) topology.algorithm = algorithm;
  emitBinary(topology, format);
}

async function handleCompress(args: string[]) {
  const input = args[0];
  let output = args[1];
  let type = 'json';

  const typeIdx = args.indexOf('--type');
  if (typeIdx !== -1 && args[typeIdx + 1]) {
    type = args[typeIdx + 1];
  }

  if (!input) {
    throw new Error('Missing input file');
  }

  if (!output) {
    output = input.endsWith('.json') ? input.replace('.json', '.bw') : `${input}.bw`;
  }

  const rawContent = fs.readFileSync(input, 'utf8');
  let data;

  if (type === 'json') {
    data = JSON.parse(rawContent);
  } else if (type === 'dna') {
    const sequence = rawContent.replace(/\s/g, '').toUpperCase();
    if (!/^[ACGT]*$/.test(sequence)) {
      throw new Error('Invalid DNA sequence');
    }
    
    // Manual bit packing for DNA (2 bits per base)
    const BASES_TO_BITS = { A: 0, C: 1, G: 2, T: 3 };
    const packedLength = Math.ceil(sequence.length / 4);
    const packedBases = new Uint8Array(packedLength);
    for (let i = 0; i < sequence.length; i++) {
      const bits = BASES_TO_BITS[sequence[i]];
      const byteIndex = Math.floor(i / 4);
      const bitPos = (3 - (i % 4)) * 2;
      packedBases[byteIndex] |= (bits << bitPos);
    }

    data = {
      header: path.basename(input),
      sequenceLength: sequence.length,
      packedBases,
    };
  } else if (type === 'gg') {
    const { BettyCompiler } = await import('../src/betty/compiler.js');
    const { compileToFlowStream } = await import('../src/packed-knot-to-flowframe.js');
    const { syzygyFoldBinary } = await import('../src/bitwise-cipher.js');

    const compiler = new BettyCompiler();
    const result = compiler.parse(rawContent);

    if (!result.ast) {
      throw new Error(`Parsing failed for ${input}: ${result.diagnostics?.map((d) => d.message).join(', ')}`);
    }

    const nodes = Array.from(result.ast.nodes.values()).map((n) => ({
      id: n.id,
      labels: n.labels,
      properties: n.properties,
    }));
    const edges = result.ast.edges.map((e) => ({
      sourceIds: e.sourceIds,
      targetIds: e.targetIds,
      type: e.type,
      properties: e.properties,
    }));

    const program = { nodes, edges };
    const stream = compileToFlowStream(program);
    const folded = await syzygyFoldBinary(stream);

    const binary = new Uint8Array(4 + folded.length);
    binary[0] = (stream.length >>> 24) & 0xff;
    binary[1] = (stream.length >>> 16) & 0xff;
    binary[2] = (stream.length >>> 8) & 0xff;
    binary[3] = stream.length & 0xff;
    binary.set(folded, 4);

    fs.writeFileSync(output, binary);
    console.log(`Compressed ${input} -> ${output} (${binary.length} bytes)`);
    return;
  } else {
    throw new Error(`Unsupported type: ${type}`);
  }

  const codec = createTopologyCodec();
  const binary = codec.encode(data);
  fs.writeFileSync(output, binary);

  console.log(`Compressed ${input} -> ${output} (${binary.length} bytes)`);
}

async function handleDecompress(args: string[]) {
  const input = args[0];
  let output = args[1];

  if (!input) {
    throw new Error('Missing input file');
  }

  const binary = fs.readFileSync(input);

  // If this is a syzygy folded binary (type gg), it may not be decodable by topology-codec yet.
  // We'll try trying topology codec as default, but if it fails we might need unfold logic.
  let data;
  const codec = createTopologyCodec();
  try {
    data = codec.decode(binary);
  } catch (err) {
    data = null; // Decryption failure or non-topology codec payload
  }

  // Attempt syzygy unfold if topology decode failed
  if (!data) {
    try {
      const { syzygyUnfoldBinary } = await import('../src/bitwise-cipher.js');
      const { flowStreamToTopology } = await import('../src/packed-knot-to-flowframe.js');
      
      const originalLength = (binary[0] << 24) | (binary[1] << 16) | (binary[2] << 8) | binary[3];
      const folded = binary.slice(4);
      const unfolded = await syzygyUnfoldBinary(folded, originalLength);
      
      if (unfolded && unfolded.length > 0) {
        data = flowStreamToTopology(unfolded);
      }
    } catch {
      // Ignore errors and fall through
    }
  }

  if (!data) {
    throw new Error('Failed to decode Lexicon binary');
  }

  let finalOutput;
  if (data.packedBases instanceof Uint8Array && typeof data.sequenceLength === 'number') {
    const { packedBasesToDnaString } = await import('../../dash/src/engine/bitwise.js');
    finalOutput = packedBasesToDnaString(data.packedBases, data.sequenceLength);
  } else {
    finalOutput = JSON.stringify(data, null, 2);
  }

  if (output) {
    fs.writeFileSync(output, finalOutput);
    console.log(`Decompressed ${input} -> ${output}`);
  } else {
    console.log(finalOutput);
  }
}

async function handleInspect(args: string[]) {
  const input = args[0];
  if (!input) {
    throw new Error('Missing input file');
  }

  const binary = fs.readFileSync(input);
  const { readHeader, MAGIC } = await import('../../dash/src/engine/lexicon.js');
  const header = readHeader(binary);

  if (!header) {
    throw new Error('Not a valid Lexicon binary');
  }

  let magicName = 'UNKNOWN';
  for (const [name, bytes] of Object.entries(MAGIC)) {
    if (bytes[0] === header.magic[0] && bytes[1] === header.magic[1]) {
      magicName = name;
      break;
    }
  }

  console.log(`Inspection for ${input}:`);
  console.log(`  Magic:   ${magicName} (${header.magic.map(b => b.toString(16).padStart(2, '0')).join(' ')})`);
  console.log(`  Version: ${header.version}`);
  console.log(`  Length:  ${header.payloadLength} bytes`);
  console.log(`  CRC:     0x${header.crc.toString(16).padStart(4, '0')}`);
}

// When executed as a script via `node`
if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch(error => {
    console.error(`Error: ${error.message}`);
    process.exit(1);
  });
}
