// @a0n/gnosis/forkjoin-app/client/components/Channels/VoiceChannel.tsx
/**
 * VoiceChannel - WebRTC Voice Chat
 *
 * Muted by default. Wild when enabled.
 * Peer-to-peer voice with WebRTC.
 */

import { useState, useRef, useEffect, memo, useCallback } from '@a0n/raect';
import {
  Headphones,
  Mic,
  MicOff,
  PhoneOff,
  Volume2,
  VolumeX,
  X,
  Users,
  Circle,
  Unlock,
} from '@a0n/aeon-icons';

interface VoiceParticipant {
  id: string;
  color: string;
  isMuted: boolean;
  isSpeaking: boolean;
}

interface VoiceChannelProps {
  participants: VoiceParticipant[];
  isConnected: boolean;
  visitorId: string;
  visitorColor: string;
  onJoin: () => void;
  onLeave: () => void;
}

export const VoiceChannel = memo(function VoiceChannel({
  participants,
  isConnected,
  visitorId,
  visitorColor,
  onJoin,
  onLeave,
}: VoiceChannelProps) {
  const [isOpen, setIsOpen] = useState(false);
  const [isEnabled, setIsEnabled] = useState(false);
  const [isInChannel, setIsInChannel] = useState(false);
  const [isMuted, setIsMuted] = useState(true);
  const [isDeafened, setIsDeafened] = useState(false);
  const localStreamRef = useRef<MediaStream | null>(null);
  const audioContextRef = useRef<AudioContext | null>(null);

  // Cleanup on unmount
  useEffect(() => {
    return () => {
      if (localStreamRef.current) {
        localStreamRef.current.getTracks().forEach((track) => track.stop());
      }
      if (audioContextRef.current) {
        audioContextRef.current.close();
      }
    };
  }, []);

  const enableVoice = async () => {
    try {
      // Request microphone access
      const stream = await navigator.mediaDevices.getUserMedia({ audio: true });
      localStreamRef.current = stream;

      // Mute by default
      stream.getAudioTracks().forEach((track) => {
        track.enabled = false;
      });

      setIsEnabled(true);
    } catch (err) {
      console.error('Failed to get microphone access:', err);
    }
  };

  const joinChannel = useCallback(() => {
    if (!isEnabled) return;
    setIsInChannel(true);
    onJoin();
  }, [isEnabled, onJoin]);

  const leaveChannel = useCallback(() => {
    setIsInChannel(false);
    setIsMuted(true);

    if (localStreamRef.current) {
      localStreamRef.current.getAudioTracks().forEach((track) => {
        track.enabled = false;
      });
    }

    onLeave();
  }, [onLeave]);

  const toggleMute = useCallback(() => {
    if (!localStreamRef.current) return;

    const newMuted = !isMuted;
    localStreamRef.current.getAudioTracks().forEach((track) => {
      track.enabled = !newMuted;
    });
    setIsMuted(newMuted);
  }, [isMuted]);

  const toggleDeafen = useCallback(() => {
    setIsDeafened(!isDeafened);
    // Implementation note: implementation, this would mute incoming audio
  }, [isDeafened]);

  const inChannelCount = participants.length + (isInChannel ? 1 : 0);

  return (
    <>
      {/* Toggle Button */}
      <button
        className="channel-toggle voice-toggle"
        onClick={() => setIsOpen(!isOpen)}
        aria-label="Toggle voice channel"
        aria-expanded={isOpen}
      >
        <Headphones size={20} />
        {inChannelCount > 0 && (
          <span className="participant-badge">{inChannelCount}</span>
        )}
      </button>

      {/* Voice Panel */}
      {isOpen && (
        <div className="voice-channel" role="dialog" aria-label="Voice channel">
          <div className="channel-header">
            <h3>Voice Channel</h3>
            <span className="channel-status">
              <Circle
                size={8}
                fill={isConnected ? '#10b981' : '#ef4444'}
                color={isConnected ? '#10b981' : '#ef4444'}
              />
              {isConnected ? 'Live' : 'Offline'}
            </span>
            <button
              className="close-button"
              onClick={() => setIsOpen(false)}
              aria-label="Close voice panel"
            >
              <X size={16} />
            </button>
          </div>

          <div className="participants-container">
            {!isEnabled ? (
              <div className="enable-prompt">
                <Headphones size={32} strokeWidth={1} />
                <p>Voice chat is muted by default</p>
                <button className="enable-button" onClick={enableVoice}>
                  <Unlock size={16} />
                  Enable Voice
                </button>
              </div>
            ) : !isInChannel ? (
              <div className="join-prompt">
                <Users size={32} strokeWidth={1} />
                <p>{inChannelCount} in channel</p>
                <button className="join-button" onClick={joinChannel}>
                  <Headphones size={16} />
                  Join Voice
                </button>
              </div>
            ) : (
              <>
                {/* Self */}
                <div className="participant self">
                  <span
                    className="participant-avatar"
                    style={{ backgroundColor: visitorColor }}
                  >
                    {isMuted && (
                      <span className="muted-indicator">
                        <MicOff size={10} />
                      </span>
                    )}
                  </span>
                  <span className="participant-name">You</span>
                  <div className="participant-controls">
                    <button
                      className={`control-button ${isMuted ? 'active' : ''}`}
                      onClick={toggleMute}
                      aria-label={isMuted ? 'Unmute' : 'Mute'}
                    >
                      {isMuted ? <MicOff size={14} /> : <Mic size={14} />}
                    </button>
                    <button
                      className={`control-button ${isDeafened ? 'active' : ''}`}
                      onClick={toggleDeafen}
                      aria-label={isDeafened ? 'Undeafen' : 'Deafen'}
                    >
                      {isDeafened ? (
                        <VolumeX size={14} />
                      ) : (
                        <Volume2 size={14} />
                      )}
                    </button>
                  </div>
                </div>

                {/* Other participants */}
                {participants.map((p) => (
                  <div
                    key={p.id}
                    className={`participant ${p.isSpeaking ? 'speaking' : ''}`}
                  >
                    <span
                      className="participant-avatar"
                      style={{ backgroundColor: p.color }}
                    >
                      {p.isMuted && (
                        <span className="muted-indicator">
                          <MicOff size={10} />
                        </span>
                      )}
                    </span>
                    <span className="participant-name">
                      Visitor {p.id.slice(0, 6)}
                    </span>
                  </div>
                ))}

                {participants.length === 0 && (
                  <div className="empty-channel">
                    <p>No one else here yet</p>
                  </div>
                )}
              </>
            )}
          </div>

          {isInChannel && (
            <div className="voice-controls">
              <button className="leave-button" onClick={leaveChannel}>
                <PhoneOff size={16} />
                Leave Voice
              </button>
            </div>
          )}
        </div>
      )}

      <style>{`
        .voice-toggle {
          bottom: 136px;
          right: 24px;
        }

        .participant-badge {
          position: absolute;
          top: -4px;
          right: -4px;
          min-width: 18px;
          height: 18px;
          border-radius: 9px;
          background: #10b981;
          color: white;
          font-size: 0.7rem;
          font-weight: 600;
          display: flex;
          align-items: center;
          justify-content: center;
          padding: 0 4px;
        }

        .voice-channel {
          position: fixed;
          bottom: 80px;
          right: 80px;
          width: 280px;
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
          display: flex;
          flex-direction: column;
          z-index: 1000;
          box-shadow: 0 8px 32px rgba(0, 0, 0, 0.3);
          animation: slide-up 0.2s ease;
        }

        .participants-container {
          padding: var(--space-md);
          min-height: 150px;
        }

        .enable-prompt,
        .join-prompt {
          display: flex;
          flex-direction: column;
          align-items: center;
          justify-content: center;
          height: 150px;
          color: var(--color-text-muted);
          text-align: center;
          gap: var(--space-md);
        }

        .enable-prompt p,
        .join-prompt p {
          margin: 0;
          font-size: 0.85rem;
        }

        .join-button,
        .leave-button {
          display: flex;
          align-items: center;
          gap: var(--space-xs);
          padding: var(--space-sm) var(--space-md);
          border-radius: 8px;
          border: none;
          cursor: pointer;
          font-size: 0.85rem;
          transition: all 0.2s ease;
        }

        .join-button {
          background: #10b981;
          color: white;
        }

        .join-button:hover {
          background: #059669;
        }

        .leave-button {
          background: #ef4444;
          color: white;
        }

        .leave-button:hover {
          background: #dc2626;
        }

        .participant {
          display: flex;
          align-items: center;
          gap: var(--space-sm);
          padding: var(--space-xs) 0;
        }

        .participant.speaking .participant-avatar {
          box-shadow: 0 0 0 3px #10b981;
        }

        .participant-avatar {
          position: relative;
          width: 32px;
          height: 32px;
          border-radius: 50%;
          display: flex;
          align-items: center;
          justify-content: center;
        }

        .muted-indicator {
          position: absolute;
          bottom: -2px;
          right: -2px;
          width: 14px;
          height: 14px;
          border-radius: 50%;
          background: var(--color-bg-secondary);
          display: flex;
          align-items: center;
          justify-content: center;
          color: #ef4444;
        }

        .participant-name {
          flex: 1;
          font-size: 0.85rem;
          color: var(--color-text);
        }

        .participant-controls {
          display: flex;
          gap: 4px;
        }

        .control-button {
          width: 28px;
          height: 28px;
          border-radius: 6px;
          border: none;
          background: var(--color-bg-tertiary);
          color: var(--color-text-muted);
          cursor: pointer;
          display: flex;
          align-items: center;
          justify-content: center;
          transition: all 0.2s ease;
        }

        .control-button:hover {
          color: var(--color-text);
        }

        .control-button.active {
          background: #ef4444;
          color: white;
        }

        .empty-channel {
          text-align: center;
          color: var(--color-text-muted);
          font-size: 0.8rem;
          padding: var(--space-md);
        }

        .voice-controls {
          padding: var(--space-sm) var(--space-md);
          border-top: 1px solid var(--color-bg-tertiary);
          display: flex;
          justify-content: center;
        }

        @media (max-width: 480px) {
          .voice-channel {
            right: 12px;
            left: 12px;
            width: auto;
            bottom: 130px;
          }

          .voice-toggle {
            right: 12px;
            bottom: 124px;
          }
        }
      `}</style>
    </>
  );
});
