// @a0n/.ignored_gnosis/frffmpeg/js/frffmpeg-wasm.test.mjs
/**
 * Node integration test for the frffmpeg WASM bridge.
 *
 * Run from this directory:
 *   node frffmpeg-wasm.test.mjs
 *
 * No external dependencies — uses only Node built-ins + the bundled frffmpeg.wasm.
 * The test encodes a 4-frame 64×48 RGB gradient clip and verifies the output is a
 * structurally valid MP4 containing ftyp / moov / mdat / avcC boxes.
 */

import { readFileSync, writeFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";
import { tmpdir } from "node:os";
import { spawnSync } from "node:child_process";

// ---------------------------------------------------------------------------
// Inline the bridge (JS-only re-implementation of frffmpeg-wasm.ts so the test
// runs without a TypeScript build step, but exercises exactly the same logic).
// ---------------------------------------------------------------------------

async function loadFrffmpeg(wasmSource) {
  let instance;
  if (wasmSource instanceof WebAssembly.Module) {
    instance = await WebAssembly.instantiate(wasmSource, {});
  } else {
    const result = await WebAssembly.instantiate(wasmSource, {});
    instance = result.instance;
  }
  const exports = instance.exports;
  for (const name of ["memory", "frffmpeg_alloc", "frffmpeg_free", "frffmpeg_encode_ipcm", "frffmpeg_encode_compressed", "frffmpeg_encode_rate_controlled", "frffmpeg_encode_rate_controlled_av", "frffmpeg_encode_av", "frffmpeg_encode_sample_compressed", "frffmpeg_encode_rate_controlled_samples", "frffmpeg_mux_samples", "frffmpeg_mux_samples_av"]) {
    if (!(name in exports)) throw new Error(`frffmpeg WASM missing export: "${name}"`);
  }
  return new Frffmpeg(exports);
}

class Frffmpeg {
  constructor(exports) {
    this.exports = exports;
  }

  encodeRgbFramesToMp4(frames, width, height, fps) {
    const { exports } = this;
    const n = frames.length;
    if (n === 0) throw new RangeError("frames must be non-empty");

    const frameLen = width * height * 3;
    for (let i = 0; i < n; i++) {
      if (frames[i].length !== frameLen) {
        throw new RangeError(
          `frame[${i}] has ${frames[i].length} bytes; expected ${frameLen}`
        );
      }
    }
    const totalLen = frameLen * n;

    // alloc input
    const inPtr = exports.frffmpeg_alloc(totalLen);
    // Re-acquire after alloc (memory may have grown).
    let mem = new Uint8Array(exports.memory.buffer);
    let off = inPtr;
    for (let i = 0; i < n; i++) {
      mem.set(frames[i], off);
      off += frameLen;
    }

    // alloc 4-byte out_len cell
    const outLenPtr = exports.frffmpeg_alloc(4);
    // Re-acquire after second alloc.
    mem = new Uint8Array(exports.memory.buffer);
    mem[outLenPtr] = mem[outLenPtr + 1] = mem[outLenPtr + 2] = mem[outLenPtr + 3] = 0;

    // encode
    const outPtr = exports.frffmpeg_encode_ipcm(
      inPtr, frameLen, n, width, height, fps, outLenPtr
    );

    // Re-acquire AFTER encode (internal allocs can grow memory).
    const outLen = new DataView(exports.memory.buffer).getUint32(outLenPtr, true);

    let result;
    if (outPtr === 0 || outLen === 0) {
      result = new Uint8Array(0);
    } else {
      result = new Uint8Array(exports.memory.buffer, outPtr, outLen).slice();
      exports.frffmpeg_free(outPtr, outLen);
    }

    exports.frffmpeg_free(inPtr, totalLen);
    exports.frffmpeg_free(outLenPtr, 4);

    return result;
  }

  encodeRgbFramesToMp4Compressed(frames, width, height, fps, qp = 26) {
    const { exports } = this;
    const n = frames.length;
    if (n === 0) throw new RangeError("frames must be non-empty");

    const frameLen = width * height * 3;
    for (let i = 0; i < n; i++) {
      if (frames[i].length !== frameLen) {
        throw new RangeError(
          `frame[${i}] has ${frames[i].length} bytes; expected ${frameLen}`
        );
      }
    }
    const totalLen = frameLen * n;

    // alloc input
    const inPtr = exports.frffmpeg_alloc(totalLen);
    // Re-acquire after alloc (memory may have grown).
    let mem = new Uint8Array(exports.memory.buffer);
    let off = inPtr;
    for (let i = 0; i < n; i++) {
      mem.set(frames[i], off);
      off += frameLen;
    }

    // alloc 4-byte out_len cell
    const outLenPtr = exports.frffmpeg_alloc(4);
    // Re-acquire after second alloc.
    mem = new Uint8Array(exports.memory.buffer);
    mem[outLenPtr] = mem[outLenPtr + 1] = mem[outLenPtr + 2] = mem[outLenPtr + 3] = 0;

    // encode
    const outPtr = exports.frffmpeg_encode_compressed(
      inPtr, frameLen, n, width, height, fps, qp, outLenPtr
    );

    // Re-acquire AFTER encode (internal allocs can grow memory).
    const outLen = new DataView(exports.memory.buffer).getUint32(outLenPtr, true);

    let result;
    if (outPtr === 0 || outLen === 0) {
      result = new Uint8Array(0);
    } else {
      result = new Uint8Array(exports.memory.buffer, outPtr, outLen).slice();
      exports.frffmpeg_free(outPtr, outLen);
    }

    exports.frffmpeg_free(inPtr, totalLen);
    exports.frffmpeg_free(outLenPtr, 4);

    return result;
  }

  // Rate-controlled: per-frame QP adapts to hit targetKbps (target video bitrate).
  encodeRgbFramesToMp4RateControlled(frames, width, height, fps, targetKbps) {
    const { exports } = this;
    const n = frames.length;
    if (n === 0) throw new RangeError("frames must be non-empty");

    const frameLen = width * height * 3;
    for (let i = 0; i < n; i++) {
      if (frames[i].length !== frameLen) {
        throw new RangeError(`frame[${i}] has ${frames[i].length} bytes; expected ${frameLen}`);
      }
    }
    const totalLen = frameLen * n;

    const inPtr = exports.frffmpeg_alloc(totalLen);
    let mem = new Uint8Array(exports.memory.buffer);
    let off = inPtr;
    for (let i = 0; i < n; i++) {
      mem.set(frames[i], off);
      off += frameLen;
    }

    const outLenPtr = exports.frffmpeg_alloc(4);
    mem = new Uint8Array(exports.memory.buffer);
    mem[outLenPtr] = mem[outLenPtr + 1] = mem[outLenPtr + 2] = mem[outLenPtr + 3] = 0;

    const outPtr = exports.frffmpeg_encode_rate_controlled(
      inPtr, frameLen, n, width, height, fps, targetKbps, outLenPtr
    );

    const outLen = new DataView(exports.memory.buffer).getUint32(outLenPtr, true);

    let result;
    if (outPtr === 0 || outLen === 0) {
      result = new Uint8Array(0);
    } else {
      result = new Uint8Array(exports.memory.buffer, outPtr, outLen).slice();
      exports.frffmpeg_free(outPtr, outLen);
    }

    exports.frffmpeg_free(inPtr, totalLen);
    exports.frffmpeg_free(outLenPtr, 4);

    return result;
  }

  // A/V: compressed video + AAC-LC audio → one faststart MP4 with two tracks.
  encodeRgbFramesWithAudioToMp4(frames, width, height, fps, pcm, sampleRate, channels, qp = 26) {
    const { exports } = this;
    const n = frames.length;
    if (n === 0) throw new RangeError("frames must be non-empty");

    const frameLen = width * height * 3;
    for (let i = 0; i < n; i++) {
      if (frames[i].length !== frameLen) {
        throw new RangeError(`frame[${i}] has ${frames[i].length} bytes; expected ${frameLen}`);
      }
    }
    const totalLen = frameLen * n;
    const pcmSamples = pcm.length;
    const pcmBytes = pcmSamples * 2;

    // alloc video input
    const inPtr = exports.frffmpeg_alloc(totalLen);
    let mem = new Uint8Array(exports.memory.buffer);
    let off = inPtr;
    for (let i = 0; i < n; i++) {
      mem.set(frames[i], off);
      off += frameLen;
    }

    // alloc PCM buffer and write i16 little-endian
    const pcmPtr = exports.frffmpeg_alloc(pcmBytes > 0 ? pcmBytes : 1);
    let dv = new DataView(exports.memory.buffer);
    for (let i = 0; i < pcmSamples; i++) dv.setInt16(pcmPtr + i * 2, pcm[i], true);

    // alloc 4-byte out_len cell
    const outLenPtr = exports.frffmpeg_alloc(4);
    mem = new Uint8Array(exports.memory.buffer);
    mem[outLenPtr] = mem[outLenPtr + 1] = mem[outLenPtr + 2] = mem[outLenPtr + 3] = 0;

    const outPtr = exports.frffmpeg_encode_av(
      inPtr, frameLen, n, width, height, fps, qp, pcmPtr, pcmSamples, sampleRate, channels, outLenPtr
    );

    const outLen = new DataView(exports.memory.buffer).getUint32(outLenPtr, true);

    let result;
    if (outPtr === 0 || outLen === 0) {
      result = new Uint8Array(0);
    } else {
      result = new Uint8Array(exports.memory.buffer, outPtr, outLen).slice();
      exports.frffmpeg_free(outPtr, outLen);
    }

    exports.frffmpeg_free(inPtr, totalLen);
    exports.frffmpeg_free(pcmPtr, pcmBytes > 0 ? pcmBytes : 1);
    exports.frffmpeg_free(outLenPtr, 4);

    return result;
  }

  // SHARD: encode each frame independently into a compressed AVCC sample.
  encodeFramesToSamplesCompressed(frames, width, height, qp = 26, startIdrId = 0) {
    const { exports } = this;
    const n = frames.length;
    if (n === 0) throw new RangeError("frames must be non-empty");

    const frameLen = width * height * 3;
    for (let i = 0; i < n; i++) {
      if (frames[i].length !== frameLen) {
        throw new RangeError(`frame[${i}] has ${frames[i].length} bytes; expected ${frameLen}`);
      }
    }

    const samples = new Array(n);
    for (let i = 0; i < n; i++) {
      const inPtr = exports.frffmpeg_alloc(frameLen);
      // Re-acquire after alloc (memory may have grown).
      let mem = new Uint8Array(exports.memory.buffer);
      mem.set(frames[i], inPtr);

      const outLenPtr = exports.frffmpeg_alloc(4);
      // Re-acquire after second alloc.
      mem = new Uint8Array(exports.memory.buffer);
      mem[outLenPtr] = mem[outLenPtr + 1] = mem[outLenPtr + 2] = mem[outLenPtr + 3] = 0;

      const outPtr = exports.frffmpeg_encode_sample_compressed(
        inPtr, frameLen, width, height, qp, startIdrId + i, outLenPtr
      );

      // Re-acquire AFTER encode (internal allocs can grow memory).
      const outLen = new DataView(exports.memory.buffer).getUint32(outLenPtr, true);

      if (outPtr === 0 || outLen === 0) {
        samples[i] = new Uint8Array(0);
      } else {
        samples[i] = new Uint8Array(exports.memory.buffer, outPtr, outLen).slice();
        exports.frffmpeg_free(outPtr, outLen);
      }

      exports.frffmpeg_free(inPtr, frameLen);
      exports.frffmpeg_free(outLenPtr, 4);
    }

    return samples;
  }

  // SHARD (rate-controlled): encode frames → per-frame compressed samples whose QP adapts to
  // targetKbps. Returns the SAME shape as encodeFramesToSamplesCompressed (one sample/frame,
  // in order). The controller is a sequential feedback loop, so the whole range encodes in one
  // wasm call; the wasm fills the lens table and returns the concatenated blob, which we slice.
  encodeFramesToSamplesCompressedRateControlled(frames, width, height, fps, targetKbps, startIdrId = 0) {
    void startIdrId;
    const { exports } = this;
    const n = frames.length;
    if (n === 0) throw new RangeError("frames must be non-empty");

    const frameLen = width * height * 3;
    for (let i = 0; i < n; i++) {
      if (frames[i].length !== frameLen) {
        throw new RangeError(`frame[${i}] has ${frames[i].length} bytes; expected ${frameLen}`);
      }
    }
    const totalLen = frameLen * n;

    const inPtr = exports.frffmpeg_alloc(totalLen);
    // Re-acquire after alloc.
    let mem = new Uint8Array(exports.memory.buffer);
    let offset = inPtr;
    for (let i = 0; i < n; i++) {
      mem.set(frames[i], offset);
      offset += frameLen;
    }

    const lensPtr = exports.frffmpeg_alloc(n * 4);
    // Re-acquire after second alloc.
    let dv = new DataView(exports.memory.buffer);
    for (let i = 0; i < n; i++) dv.setUint32(lensPtr + i * 4, 0, true);

    const outLenPtr = exports.frffmpeg_alloc(4);
    // Re-acquire after third alloc.
    mem = new Uint8Array(exports.memory.buffer);
    mem[outLenPtr] = mem[outLenPtr + 1] = mem[outLenPtr + 2] = mem[outLenPtr + 3] = 0;

    const outPtr = exports.frffmpeg_encode_rate_controlled_samples(
      inPtr, frameLen, n, width, height, fps, targetKbps, lensPtr, outLenPtr
    );

    // Re-acquire AFTER encode (internal allocs can grow memory).
    const outLen = new DataView(exports.memory.buffer).getUint32(outLenPtr, true);

    const samples = new Array(n);
    if (outPtr === 0 || outLen === 0) {
      for (let i = 0; i < n; i++) samples[i] = new Uint8Array(0);
    } else {
      dv = new DataView(exports.memory.buffer);
      let off = outPtr;
      for (let i = 0; i < n; i++) {
        const l = dv.getUint32(lensPtr + i * 4, true);
        samples[i] = new Uint8Array(exports.memory.buffer, off, l).slice();
        off += l;
      }
      exports.frffmpeg_free(outPtr, outLen);
    }

    exports.frffmpeg_free(inPtr, totalLen);
    exports.frffmpeg_free(lensPtr, n * 4);
    exports.frffmpeg_free(outLenPtr, 4);

    return samples;
  }

  // FOLD: mux pre-encoded shard samples (in frame order) into one MP4.
  muxSamplesToMp4(samples, width, height, fps) {
    const { exports } = this;
    const n = samples.length;
    if (n === 0) throw new RangeError("samples must be non-empty");

    let totalLen = 0;
    for (let i = 0; i < n; i++) totalLen += samples[i].length;

    const samplesPtr = exports.frffmpeg_alloc(totalLen);
    // Re-acquire after alloc.
    let mem = new Uint8Array(exports.memory.buffer);
    let off = samplesPtr;
    for (let i = 0; i < n; i++) {
      mem.set(samples[i], off);
      off += samples[i].length;
    }

    const lensPtr = exports.frffmpeg_alloc(n * 4);
    // Re-acquire after second alloc.
    let dv = new DataView(exports.memory.buffer);
    for (let i = 0; i < n; i++) dv.setUint32(lensPtr + i * 4, samples[i].length, true);

    const outLenPtr = exports.frffmpeg_alloc(4);
    // Re-acquire after third alloc.
    mem = new Uint8Array(exports.memory.buffer);
    mem[outLenPtr] = mem[outLenPtr + 1] = mem[outLenPtr + 2] = mem[outLenPtr + 3] = 0;

    const outPtr = exports.frffmpeg_mux_samples(
      samplesPtr, lensPtr, n, width, height, fps, outLenPtr
    );

    // Re-acquire AFTER mux (internal allocs can grow memory).
    const outLen = new DataView(exports.memory.buffer).getUint32(outLenPtr, true);

    let result;
    if (outPtr === 0 || outLen === 0) {
      result = new Uint8Array(0);
    } else {
      result = new Uint8Array(exports.memory.buffer, outPtr, outLen).slice();
      exports.frffmpeg_free(outPtr, outLen);
    }

    exports.frffmpeg_free(samplesPtr, totalLen);
    exports.frffmpeg_free(lensPtr, n * 4);
    exports.frffmpeg_free(outLenPtr, 4);

    return result;
  }

  // FOLD WITH AUDIO: mux pre-encoded shard video samples + i16 PCM → two-track MP4.
  muxSamplesToMp4WithAudio(samples, width, height, fps, pcm, sampleRate, channels) {
    const { exports } = this;
    const n = samples.length;
    if (n === 0) throw new RangeError("samples must be non-empty");

    let totalLen = 0;
    for (let i = 0; i < n; i++) totalLen += samples[i].length;

    const pcmSamples = pcm.length;
    const pcmBytes = pcmSamples * 2;

    const samplesPtr = exports.frffmpeg_alloc(totalLen);
    // Re-acquire after alloc.
    let mem = new Uint8Array(exports.memory.buffer);
    let off = samplesPtr;
    for (let i = 0; i < n; i++) {
      mem.set(samples[i], off);
      off += samples[i].length;
    }

    const lensPtr = exports.frffmpeg_alloc(n * 4);
    // Re-acquire after second alloc.
    let dv = new DataView(exports.memory.buffer);
    for (let i = 0; i < n; i++) dv.setUint32(lensPtr + i * 4, samples[i].length, true);

    const pcmPtr = exports.frffmpeg_alloc(pcmBytes > 0 ? pcmBytes : 1);
    // Re-acquire after alloc; write i16 little-endian.
    dv = new DataView(exports.memory.buffer);
    for (let i = 0; i < pcmSamples; i++) dv.setInt16(pcmPtr + i * 2, pcm[i], true);

    const outLenPtr = exports.frffmpeg_alloc(4);
    // Re-acquire after alloc.
    mem = new Uint8Array(exports.memory.buffer);
    mem[outLenPtr] = mem[outLenPtr + 1] = mem[outLenPtr + 2] = mem[outLenPtr + 3] = 0;

    const outPtr = exports.frffmpeg_mux_samples_av(
      samplesPtr, lensPtr, n, width, height, fps, pcmPtr, pcmSamples, sampleRate, channels, outLenPtr
    );

    // Re-acquire AFTER mux (internal allocs can grow memory).
    const outLen = new DataView(exports.memory.buffer).getUint32(outLenPtr, true);

    let result;
    if (outPtr === 0 || outLen === 0) {
      result = new Uint8Array(0);
    } else {
      result = new Uint8Array(exports.memory.buffer, outPtr, outLen).slice();
      exports.frffmpeg_free(outPtr, outLen);
    }

    exports.frffmpeg_free(samplesPtr, totalLen);
    exports.frffmpeg_free(lensPtr, n * 4);
    exports.frffmpeg_free(pcmPtr, pcmBytes > 0 ? pcmBytes : 1);
    exports.frffmpeg_free(outLenPtr, 4);

    return result;
  }
}

// ---------------------------------------------------------------------------
// Test helpers
// ---------------------------------------------------------------------------

let passed = 0;
let failed = 0;

function assert(cond, msg) {
  if (cond) {
    console.log(`  PASS  ${msg}`);
    passed++;
  } else {
    console.error(`  FAIL  ${msg}`);
    failed++;
  }
}

function boxPresent(mp4, tag) {
  const needle = Buffer.from(tag, "ascii");
  return Buffer.from(mp4.buffer, mp4.byteOffset, mp4.byteLength).includes(needle);
}

/**
 * Build a test RGB24 frame: horizontal red→blue gradient with a small white square
 * that moves across frames so they differ from each other.
 */
function makeTestFrame(w, h, t) {
  const rgb = new Uint8Array(w * h * 3);
  for (let y = 0; y < h; y++) {
    for (let x = 0; x < w; x++) {
      const i = (y * w + x) * 3;
      const g = Math.floor((255 * x) / w);
      rgb[i]     = 255 - g; // R fades L→R
      rgb[i + 1] = 0;
      rgb[i + 2] = g;       // B grows L→R
      // moving white square
      const sx = (4 + t * 3) % w;
      if (x >= sx && x < sx + 8 && y >= 8 && y < 16) {
        rgb[i] = rgb[i + 1] = rgb[i + 2] = 255;
      }
    }
  }
  return rgb;
}

// ---------------------------------------------------------------------------
// Main test
// ---------------------------------------------------------------------------

async function main() {
  const __dirname = dirname(fileURLToPath(import.meta.url));
  const wasmPath = join(__dirname, "frffmpeg.wasm");

  console.log("frffmpeg-wasm integration test");
  console.log("================================");
  console.log(`wasm: ${wasmPath}`);

  // 1. Load the wasm
  const wasmBytes = readFileSync(wasmPath);
  console.log(`wasm size: ${wasmBytes.length} bytes`);
  assert(wasmBytes.length > 0, "wasm file is non-empty");

  const enc = await loadFrffmpeg(wasmBytes);
  assert(typeof enc.encodeRgbFramesToMp4 === "function", "loadFrffmpeg returns encoder with encodeRgbFramesToMp4");

  // 2. Build a 4-frame 64×48 gradient clip
  const W = 64, H = 48, FPS = 12, N = 4;
  const frames = Array.from({ length: N }, (_, t) => makeTestFrame(W, H, t));
  console.log(`\nEncoding ${N} frames at ${W}×${H} @ ${FPS}fps...`);

  const mp4 = enc.encodeRgbFramesToMp4(frames, W, H, FPS);
  console.log(`output: ${mp4.length} bytes`);

  // 3. Structural MP4 assertions
  assert(mp4.length > 0, "output is non-empty");
  assert(mp4.length > frames[0].length, "output larger than one raw frame (plausible size)");
  assert(boxPresent(mp4, "ftyp"), 'MP4 contains "ftyp" box');
  assert(boxPresent(mp4, "moov"), 'MP4 contains "moov" box');
  assert(boxPresent(mp4, "mdat"), 'MP4 contains "mdat" box');
  assert(boxPresent(mp4, "avcC"), 'MP4 contains "avcC" decoder config');

  // 4. Verify returned buffer is an independent copy (not a wasm memory view)
  const originalByte = mp4[0];
  // If mp4 were a live view into wasm memory, a subsequent alloc could mutate it.
  enc.encodeRgbFramesToMp4([makeTestFrame(W, H, 99)], W, H, FPS); // second encode
  assert(mp4[0] === originalByte, "returned Uint8Array is a copied, stable buffer");

  // 5. Error path: wrong frame size
  let threw = false;
  try {
    enc.encodeRgbFramesToMp4([new Uint8Array(3)], W, H, FPS);
  } catch (e) {
    threw = true;
  }
  assert(threw, "throws RangeError for wrong frame size");

  // 6. Compressed encoder: structural MP4 + substantially smaller than I_PCM
  assert(
    typeof enc.encodeRgbFramesToMp4Compressed === "function",
    "encoder exposes encodeRgbFramesToMp4Compressed",
  );
  const compressedFrames = Array.from({ length: N }, (_, t) => makeTestFrame(W, H, t));
  const compressed = enc.encodeRgbFramesToMp4Compressed(compressedFrames, W, H, FPS, 26);
  console.log(`\ncompressed output: ${compressed.length} bytes`);

  assert(compressed.length > 0, "compressed output is non-empty");
  assert(boxPresent(compressed, "ftyp"), 'compressed MP4 contains "ftyp" box');
  assert(boxPresent(compressed, "moov"), 'compressed MP4 contains "moov" box');
  assert(boxPresent(compressed, "mdat"), 'compressed MP4 contains "mdat" box');
  assert(boxPresent(compressed, "avcC"), 'compressed MP4 contains "avcC" decoder config');

  // Re-encode I_PCM of the SAME clip for an apples-to-apples size comparison.
  const ipcm = enc.encodeRgbFramesToMp4(compressedFrames, W, H, FPS);
  const ratio = ipcm.length / compressed.length;
  console.log(`I_PCM output:      ${ipcm.length} bytes  (compressed is ${ratio.toFixed(1)}× smaller)`);
  assert(
    compressed.length * 2 < ipcm.length,
    `compressed (${compressed.length}) is substantially smaller than I_PCM (${ipcm.length})`,
  );

  // Write the compressed MP4 so it can be ffmpeg-decoded externally.
  const outPath = join(tmpdir(), "frffmpeg_wasm_compressed.mp4");
  writeFileSync(outPath, compressed);
  console.log(`wrote compressed MP4 → ${outPath}`);

  // 7. SHARD == WHOLE proof: fanout into independent frame-ranges, fold into one MP4.
  console.log(`\nSHARD==WHOLE: fanout 8-frame clip into [0,4)+[4,8), fold into one MP4...`);
  assert(
    typeof enc.encodeFramesToSamplesCompressed === "function",
    "encoder exposes encodeFramesToSamplesCompressed (shard primitive)",
  );
  assert(
    typeof enc.muxSamplesToMp4 === "function",
    "encoder exposes muxSamplesToMp4 (fold primitive)",
  );

  const SW = 64, SH = 48, SFPS = 12, SN = 8, SQP = 26;
  const shardFrames = Array.from({ length: SN }, (_, t) => makeTestFrame(SW, SH, t));

  // (a) WHOLE path: one-shot compressed MP4 + the whole-clip samples for byte comparison.
  const wholeMp4 = enc.encodeRgbFramesToMp4Compressed(shardFrames, SW, SH, SFPS, SQP);
  const wholeSamples = enc.encodeFramesToSamplesCompressed(shardFrames, SW, SH, SQP, 0);

  // (b) SHARDED path: two INDEPENDENT fanout encodes, each with its own startIdrId offset.
  const shardA = enc.encodeFramesToSamplesCompressed(shardFrames.slice(0, 4), SW, SH, SQP, 0);
  const shardB = enc.encodeFramesToSamplesCompressed(shardFrames.slice(4, 8), SW, SH, SQP, 4);
  const allSamples = [...shardA, ...shardB];
  const shardedMp4 = enc.muxSamplesToMp4(allSamples, SW, SH, SFPS);

  // Both are valid, plausibly-sized MP4s.
  assert(boxPresent(wholeMp4, "ftyp") && boxPresent(wholeMp4, "moov") && boxPresent(wholeMp4, "mdat"), "whole MP4 has ftyp/moov/mdat");
  assert(boxPresent(shardedMp4, "ftyp") && boxPresent(shardedMp4, "moov") && boxPresent(shardedMp4, "mdat"), "sharded MP4 has ftyp/moov/mdat");
  const payload = allSamples.reduce((s, a) => s + a.length, 0);
  assert(shardedMp4.length > payload && shardedMp4.length > 256, "sharded MP4 is plausibly sized (> sample payload)");

  // Per-frame independence: each shard sample equals the whole-clip sample for that frame.
  assert(allSamples.length === wholeSamples.length, "shard sample count equals whole-clip count");
  let samplesEqual = allSamples.length === wholeSamples.length;
  for (let i = 0; i < allSamples.length && samplesEqual; i++) {
    if (allSamples[i].length !== wholeSamples[i].length) { samplesEqual = false; break; }
    for (let j = 0; j < allSamples[i].length; j++) {
      if (allSamples[i][j] !== wholeSamples[i][j]) { samplesEqual = false; break; }
    }
  }
  assert(samplesEqual, "SHARD==WHOLE: sharded per-frame samples are byte-identical to whole-clip samples");

  // The fold is deterministic over identical samples → identical MP4 to the one-shot encode.
  let mp4Equal = shardedMp4.length === wholeMp4.length;
  for (let i = 0; i < shardedMp4.length && mp4Equal; i++) {
    if (shardedMp4[i] !== wholeMp4[i]) { mp4Equal = false; break; }
  }
  assert(mp4Equal, "SHARD==WHOLE: folded MP4 is byte-identical to the one-shot whole-clip MP4");

  // Drop the sharded artifact for external ffmpeg-decode verification.
  const shardedPath = join(tmpdir(), "frffmpeg_sharded.mp4");
  writeFileSync(shardedPath, shardedMp4);
  console.log(`wrote sharded MP4 → ${shardedPath} (${shardedMp4.length} bytes)`);

  // 8b. RATE CONTROL: per-frame QP adapts to hit a target video bitrate.
  console.log(`\nRATE CONTROL: 640×360 @24fps screen-like content, targets 800 vs 1500 kbps...`);
  assert(
    typeof enc.encodeRgbFramesToMp4RateControlled === "function",
    "encoder exposes encodeRgbFramesToMp4RateControlled",
  );

  // Broadcaster-like content: smooth panels + a scrolling ticker + a moving blob.
  const screenFrame = (w, h, t) => {
    const rgb = new Uint8Array(w * h * 3);
    for (let y = 0; y < h; y++) {
      for (let x = 0; x < w; x++) {
        const i = (y * w + x) * 3;
        const base = Math.floor(40 + 30 * (y / h));
        const tick = y > (h * 5) / 6 ? ((x + t * 6) % 40 < 6 ? 200 : 20) : 0;
        const bx = Math.floor(w * (0.5 + 0.3 * Math.sin(t * 0.3)));
        const blob =
          Math.abs(x - bx) < 24 && Math.abs(y - Math.floor(h / 3)) < 24 ? 120 : 0;
        rgb[i] = Math.min(255, base + tick + blob);
        rgb[i + 1] = Math.min(255, base + (tick >> 1) + blob);
        rgb[i + 2] = Math.min(255, base + 20 + (blob >> 1));
      }
    }
    return rgb;
  };

  const RCW = 640, RCH = 360, RCFPS = 24, RCN = 48; // 2s
  const rcFrames = Array.from({ length: RCN }, (_, t) => screenFrame(RCW, RCH, t));
  const rc800 = enc.encodeRgbFramesToMp4RateControlled(rcFrames, RCW, RCH, RCFPS, 800);
  const rc1500 = enc.encodeRgbFramesToMp4RateControlled(rcFrames, RCW, RCH, RCFPS, 1500);
  console.log(`rate-controlled: 800kbps→${rc800.length} bytes, 1500kbps→${rc1500.length} bytes`);

  assert(rc800.length > 0 && rc1500.length > 0, "rate-controlled outputs are non-empty");
  assert(boxPresent(rc800, "ftyp") && boxPresent(rc800, "moov") && boxPresent(rc800, "mdat"), "rc800 MP4 has ftyp/moov/mdat");
  assert(rc800.length < rc1500.length, "lower target (800k) is strictly smaller than higher (1500k)");

  const rcHasBin = (bin) => {
    const r = spawnSync(bin, ["-version"], { stdio: "ignore" });
    return !r.error && r.status === 0;
  };
  for (const [tag, buf] of [["800", rc800], ["1500", rc1500]]) {
    const p = join(tmpdir(), `frffmpeg_rc_${tag}.mp4`);
    writeFileSync(p, buf);
    if (rcHasBin("ffmpeg")) {
      const dec = spawnSync("ffmpeg", ["-v", "error", "-i", p, "-f", "null", "-"], { encoding: "utf8" });
      const stderr = (dec.stderr || "").trim();
      if (stderr) console.log(`rc ${tag}k ffmpeg stderr:\n${stderr}`);
      assert(dec.status === 0 && stderr === "", `rate-controlled ${tag}k decodes clean (ffmpeg exit 0, no errors)`);
    }
    if (rcHasBin("ffprobe")) {
      const pr = spawnSync("ffprobe", ["-v", "error", "-show_entries", "format=bit_rate,size", "-of", "default=nw=1", p], { encoding: "utf8" });
      console.log(`rc ${tag}k ffprobe: ${(pr.stdout || "").trim().replace(/\n/g, " ")}`);
    }
  }

  // 8c. RATE-CONTROLLED SAMPLES (the SHARDED path): encode rate-controlled per-frame samples,
  // fold via muxSamplesToMp4 (exactly the broadcaster DO fold), decode-verify, and compare
  // ffprobe sizes between an 800k and a 1500k target.
  console.log(`\nRATE-CONTROLLED SAMPLES: 640×360 @24fps → encodeFramesToSamplesCompressedRateControlled → muxSamplesToMp4...`);
  assert(
    typeof enc.encodeFramesToSamplesCompressedRateControlled === "function",
    "encoder exposes encodeFramesToSamplesCompressedRateControlled (sharded RC primitive)",
  );

  const rcs800 = enc.encodeFramesToSamplesCompressedRateControlled(rcFrames, RCW, RCH, RCFPS, 800);
  const rcs1500 = enc.encodeFramesToSamplesCompressedRateControlled(rcFrames, RCW, RCH, RCFPS, 1500);

  // SHAPE PARITY: identical to encodeFramesToSamplesCompressed — one sample buffer per input
  // frame, in order (so the DO fold + mux is unaffected).
  const fixedSamples = enc.encodeFramesToSamplesCompressed(rcFrames, RCW, RCH, 26, 0);
  assert(Array.isArray(rcs800), "RC samples is an array");
  assert(rcs800.length === rcFrames.length, `RC samples: one per frame (got ${rcs800.length}, expected ${rcFrames.length})`);
  assert(rcs1500.length === rcFrames.length, "RC samples (1500k): one per frame");
  assert(rcs800.length === fixedSamples.length, "RC samples count matches fixed-QP samples count (same shape)");
  assert(rcs800.every((s) => s instanceof Uint8Array && s.length > 0), "every RC sample is a non-empty Uint8Array (one IDR per frame)");

  const rcs800Bytes = rcs800.reduce((s, a) => s + a.length, 0);
  const rcs1500Bytes = rcs1500.reduce((s, a) => s + a.length, 0);
  console.log(`rc-samples payload: 800k→${rcs800Bytes} bytes, 1500k→${rcs1500Bytes} bytes (${rcs800.length} samples each)`);
  assert(rcs800Bytes < rcs1500Bytes, "RC-samples: 800k payload strictly smaller than 1500k payload (rate control is honest)");

  // FOLD via the same mux the sharded DO path uses, then decode-verify + size-compare.
  const rcsMp4_800 = enc.muxSamplesToMp4(rcs800, RCW, RCH, RCFPS);
  const rcsMp4_1500 = enc.muxSamplesToMp4(rcs1500, RCW, RCH, RCFPS);
  assert(boxPresent(rcsMp4_800, "ftyp") && boxPresent(rcsMp4_800, "moov") && boxPresent(rcsMp4_800, "mdat"), "rc-samples 800k MP4 has ftyp/moov/mdat");
  assert(boxPresent(rcsMp4_1500, "ftyp") && boxPresent(rcsMp4_1500, "moov") && boxPresent(rcsMp4_1500, "mdat"), "rc-samples 1500k MP4 has ftyp/moov/mdat");
  assert(rcsMp4_800.length < rcsMp4_1500.length, "rc-samples: muxed 800k MP4 strictly smaller than 1500k MP4");

  for (const [tag, buf] of [["800", rcsMp4_800], ["1500", rcsMp4_1500]]) {
    const p = join(tmpdir(), `frffmpeg_rc_samples_${tag}.mp4`);
    writeFileSync(p, buf);
    console.log(`wrote rc-samples MP4 → ${p} (${buf.length} bytes)`);
    if (rcHasBin("ffmpeg")) {
      const dec = spawnSync("ffmpeg", ["-v", "error", "-i", p, "-f", "null", "-"], { encoding: "utf8" });
      const stderr = (dec.stderr || "").trim();
      if (stderr) console.log(`rc-samples ${tag}k ffmpeg stderr:\n${stderr}`);
      assert(dec.status === 0 && stderr === "", `rc-samples ${tag}k decodes clean (ffmpeg exit 0, no errors)`);
    }
    if (rcHasBin("ffprobe")) {
      const pr = spawnSync("ffprobe", ["-v", "error", "-show_entries", "format=bit_rate,size", "-of", "default=nw=1", p], { encoding: "utf8" });
      console.log(`rc-samples ${tag}k ffprobe: ${(pr.stdout || "").trim().replace(/\n/g, " ")}`);
    }
  }

  // 9. A/V: compressed video + AAC-LC audio → ONE faststart MP4 with TWO tracks.
  console.log(`\nA/V MUX: 8-frame 64×48 gradient + 0.7s 440Hz sine (mono 44100)...`);
  assert(
    typeof enc.encodeRgbFramesWithAudioToMp4 === "function",
    "encoder exposes encodeRgbFramesWithAudioToMp4 (A/V mux)",
  );

  const AVW = 64, AVH = 48, AVFPS = 12, AVN = 8;
  const avFrames = Array.from({ length: AVN }, (_, t) => makeTestFrame(AVW, AVH, t));

  // 0.7s of 440Hz sine, mono, 44100 Hz, i16.
  const SR = 44100, FREQ = 440, DUR = 0.7;
  const nPcm = Math.round(SR * DUR);
  const pcm = new Int16Array(nPcm);
  for (let i = 0; i < nPcm; i++) {
    pcm[i] = Math.round(Math.sin((2 * Math.PI * FREQ * i) / SR) * 0.6 * 32767);
  }

  const avMp4 = enc.encodeRgbFramesWithAudioToMp4(avFrames, AVW, AVH, AVFPS, pcm, SR, 1, 26);
  console.log(`A/V output: ${avMp4.length} bytes`);

  assert(avMp4.length > 0, "A/V output is non-empty");
  assert(boxPresent(avMp4, "ftyp"), 'A/V MP4 contains "ftyp" box');
  assert(boxPresent(avMp4, "moov"), 'A/V MP4 contains "moov" box');
  assert(boxPresent(avMp4, "mdat"), 'A/V MP4 contains "mdat" box');
  assert(boxPresent(avMp4, "avcC"), 'A/V MP4 contains "avcC" (video decoder config)');
  assert(boxPresent(avMp4, "mp4a"), 'A/V MP4 contains "mp4a" (audio sample entry)');
  assert(boxPresent(avMp4, "esds"), 'A/V MP4 contains "esds" (AAC decoder config)');

  const avPath = join(tmpdir(), "frffmpeg_av.mp4");
  writeFileSync(avPath, avMp4);
  console.log(`wrote A/V MP4 → ${avPath}`);

  // DECODE-VERIFY (the whole point): ffprobe must see TWO streams (video,h264 + audio,aac)
  // and ffmpeg must decode BOTH with no error.
  const hasBin = (bin) => {
    const r = spawnSync(bin, ["-version"], { stdio: "ignore" });
    return !r.error && r.status === 0;
  };

  if (hasBin("ffprobe")) {
    const probe = spawnSync(
      "ffprobe",
      ["-v", "error", "-show_entries", "stream=index,codec_type,codec_name", "-of", "csv", avPath],
      { encoding: "utf8" },
    );
    const out = (probe.stdout || "").trim();
    console.log(`ffprobe streams:\n${out}`);
    assert(probe.status === 0, "ffprobe parsed the A/V MP4 without error");
    // CSV rows (csv adds a 'stream' descriptor col): stream,0,h264,video / stream,1,aac,audio.
    // Order of codec_name vs codec_type is not guaranteed across ffprobe builds, so just
    // require each (codec_name, codec_type) pair to co-occur on one row.
    const rows = out.split(/\r?\n/);
    const hasVideo = rows.some((r) => r.includes("h264") && r.includes("video"));
    const hasAudio = rows.some((r) => r.includes("aac") && r.includes("audio"));
    assert(hasVideo, "ffprobe reports a video,h264 stream");
    assert(hasAudio, "ffprobe reports an audio,aac stream");
    const streamCount = out.split(/\r?\n/).filter((l) => l.startsWith("stream")).length;
    assert(streamCount === 2, `ffprobe reports exactly TWO streams (got ${streamCount})`);
  } else {
    console.log("ffprobe absent — skipping 2-stream probe");
  }

  if (hasBin("ffmpeg")) {
    const dec = spawnSync("ffmpeg", ["-v", "error", "-i", avPath, "-f", "null", "-"], {
      encoding: "utf8",
    });
    const stderr = (dec.stderr || "").trim();
    if (stderr) console.log(`ffmpeg decode stderr:\n${stderr}`);
    assert(dec.status === 0, "ffmpeg decodes the A/V MP4 (exit 0)");
    assert(stderr === "", "ffmpeg reports NO decode errors for video+audio");
  } else {
    console.log("ffmpeg absent — skipping decode-verify");
  }

  // 10. A/V FOLD: shard video samples + audio → ONE faststart MP4 with TWO tracks.
  //     This is the SHARDED (fanout) full-length clip path carrying sound.
  console.log(`\nA/V FOLD: encode 8 frames to samples + 0.7s 440Hz sine → muxSamplesToMp4WithAudio...`);
  assert(
    typeof enc.muxSamplesToMp4WithAudio === "function",
    "encoder exposes muxSamplesToMp4WithAudio (audio-carrying fold)",
  );

  const FW = 64, FH = 48, FFPS = 12, FN = 8;
  const foldFrames = Array.from({ length: FN }, (_, t) => makeTestFrame(FW, FH, t));
  // The shard/fold already produces these AVCC video samples.
  const foldSamples = enc.encodeFramesToSamplesCompressed(foldFrames, FW, FH, 26, 0);

  // 0.7s of 440Hz sine, mono, 44100 Hz, i16.
  const FSR = 44100, FFREQ = 440, FDUR = 0.7;
  const fNPcm = Math.round(FSR * FDUR);
  const foldPcm = new Int16Array(fNPcm);
  for (let i = 0; i < fNPcm; i++) {
    foldPcm[i] = Math.round(Math.sin((2 * Math.PI * FFREQ * i) / FSR) * 0.6 * 32767);
  }

  const foldAvMp4 = enc.muxSamplesToMp4WithAudio(foldSamples, FW, FH, FFPS, foldPcm, FSR, 1);
  console.log(`A/V fold output: ${foldAvMp4.length} bytes`);

  assert(foldAvMp4.length > 0, "A/V fold output is non-empty");
  assert(boxPresent(foldAvMp4, "ftyp"), 'A/V fold MP4 contains "ftyp" box');
  assert(boxPresent(foldAvMp4, "moov"), 'A/V fold MP4 contains "moov" box');
  assert(boxPresent(foldAvMp4, "mdat"), 'A/V fold MP4 contains "mdat" box');
  assert(boxPresent(foldAvMp4, "esds"), 'A/V fold MP4 contains "esds" (AAC decoder config)');
  assert(boxPresent(foldAvMp4, "mp4a"), 'A/V fold MP4 contains "mp4a" (audio sample entry)');

  const foldAvPath = join(tmpdir(), "frffmpeg_samples_av.mp4");
  writeFileSync(foldAvPath, foldAvMp4);
  console.log(`wrote A/V fold MP4 → ${foldAvPath}`);

  // DECODE-VERIFY: ffprobe must see TWO streams (h264 video + aac audio); ffmpeg decodes clean.
  if (hasBin("ffprobe")) {
    const probe = spawnSync(
      "ffprobe",
      ["-v", "error", "-show_entries", "stream=codec_type,codec_name", "-of", "csv", foldAvPath],
      { encoding: "utf8" },
    );
    const out = (probe.stdout || "").trim();
    console.log(`ffprobe streams (fold):\n${out}`);
    assert(probe.status === 0, "ffprobe parsed the A/V fold MP4 without error");
    const rows = out.split(/\r?\n/);
    assert(rows.some((r) => r.includes("h264") && r.includes("video")), "fold: ffprobe reports a video,h264 stream");
    assert(rows.some((r) => r.includes("aac") && r.includes("audio")), "fold: ffprobe reports an audio,aac stream");
    const streamCount = rows.filter((l) => l.trim().length > 0).length;
    assert(streamCount === 2, `fold: ffprobe reports exactly TWO streams (got ${streamCount})`);
  } else {
    console.log("ffprobe absent — skipping fold 2-stream probe");
  }

  if (hasBin("ffmpeg")) {
    const dec = spawnSync("ffmpeg", ["-v", "error", "-i", foldAvPath, "-f", "null", "-"], {
      encoding: "utf8",
    });
    const stderr = (dec.stderr || "").trim();
    if (stderr) console.log(`ffmpeg decode stderr (fold):\n${stderr}`);
    assert(dec.status === 0, "fold: ffmpeg decodes the A/V MP4 (exit 0)");
    assert(stderr === "", "fold: ffmpeg reports NO decode errors for video+audio");
  } else {
    console.log("ffmpeg absent — skipping fold decode-verify");
  }

  // 8. Summary
  console.log(`\n================================`);
  if (failed === 0) {
    console.log(`ALL ${passed} assertions passed.`);
  } else {
    console.error(`${failed} FAILED / ${passed} passed.`);
    process.exit(1);
  }
}

main().catch((e) => {
  console.error("Unhandled error:", e);
  process.exit(1);
});
