// @a0n/.ignored_gnosis/distributed-inference/pkg-test-node/distributed_inference.js
/* @ts-self-types="./distributed_inference.d.ts" */

/**
 * JS-facing builder for streaming tensor bytes directly into WASM linear
 * memory.
 *
 * The naive path (`backend.set_tensor(name, new Uint8Array(megabytes))`)
 * requires JS to materialize the entire tensor as one contiguous
 * Uint8Array before crossing the boundary. For qwen-7b's FFN matrices
 * (~14 MB Q4_K each) this is fine on a laptop but blows the JS heap on
 * a 128 MB CF Worker, especially because the Worker is *also* holding
 * the previous phase's tensors awaiting eviction.
 *
 * `TensorBuilder` lets the host stream the row in chunks: allocate a
 * fixed-capacity Vec on the WASM side, append each chunk as it arrives
 * from R2 (Range responses are themselves chunked), and finalize by
 * handing the Vec to the backend in a zero-copy move. Peak JS heap
 * drops to one chunk's worth of bytes (kilobytes, not megabytes).
 */
class TensorBuilder {
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        TensorBuilderFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_tensorbuilder_free(ptr, 0);
    }
    /**
     * Append a chunk of bytes to the in-flight buffer. Safe to call
     * many times; the underlying Vec grows in amortized O(1) when the
     * initial capacity was right.
     * @param {Uint8Array} chunk
     */
    append(chunk) {
        const ptr0 = passArray8ToWasm0(chunk, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        wasm.tensorbuilder_append(this.__wbg_ptr, ptr0, len0);
    }
    /**
     * Hand the accumulated bytes to the backend under `name` and
     * consume the builder. After this call the builder is unusable.
     * The Vec move is zero-copy — no extra allocation crosses into
     * the backend.
     * @param {WasmHostBackend} backend
     * @param {string} name
     */
    finish(backend, name) {
        const ptr = this.__destroy_into_raw();
        _assertClass(backend, WasmHostBackend);
        const ptr0 = passStringToWasm0(name, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        wasm.tensorbuilder_finish(ptr, backend.__wbg_ptr, ptr0, len0);
    }
    /**
     * Pre-allocate `capacity` bytes. Caller is expected to pass the
     * exact final size (known from the knot metadata `length` field);
     * passing a smaller value forces re-allocations on each `append`,
     * passing a larger value wastes WASM memory. Either is correct,
     * but exact is what the worker should pass.
     * @param {number} capacity
     */
    constructor(capacity) {
        const ret = wasm.tensorbuilder_new(capacity);
        this.__wbg_ptr = ret >>> 0;
        TensorBuilderFinalization.register(this, this.__wbg_ptr, this);
        return this;
    }
}
if (Symbol.dispose) TensorBuilder.prototype[Symbol.dispose] = TensorBuilder.prototype.free;
exports.TensorBuilder = TensorBuilder;

class WasmGemma4Pipeline {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(WasmGemma4Pipeline.prototype);
        obj.__wbg_ptr = ptr;
        WasmGemma4PipelineFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        WasmGemma4PipelineFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_wasmgemma4pipeline_free(ptr, 0);
    }
    /**
     * @param {Float32Array} input
     * @param {number} layer
     * @param {number} start_pos
     * @param {number} seq_len
     * @param {boolean} is_first_layer
     */
    forward_range_attn_chunk(input, layer, start_pos, seq_len, is_first_layer) {
        const ptr0 = passArrayF32ToWasm0(input, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.wasmgemma4pipeline_forward_range_attn_chunk(this.__wbg_ptr, ptr0, len0, layer, start_pos, seq_len, is_first_layer);
        if (ret[1]) {
            throw takeFromExternrefTable0(ret[0]);
        }
    }
    /**
     * @param {number} layer
     * @param {number} seq_len
     * @param {boolean} is_last_layer
     * @returns {Float32Array}
     */
    forward_range_ffn_down_chunk(layer, seq_len, is_last_layer) {
        const ret = wasm.wasmgemma4pipeline_forward_range_ffn_down_chunk(this.__wbg_ptr, layer, seq_len, is_last_layer);
        if (ret[3]) {
            throw takeFromExternrefTable0(ret[2]);
        }
        var v1 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * @param {number} layer
     * @param {number} seq_len
     */
    forward_range_ffn_gate_chunk(layer, seq_len) {
        const ret = wasm.wasmgemma4pipeline_forward_range_ffn_gate_chunk(this.__wbg_ptr, layer, seq_len);
        if (ret[1]) {
            throw takeFromExternrefTable0(ret[0]);
        }
    }
    /**
     * @param {number} layer
     * @param {number} seq_len
     */
    forward_range_ffn_up_chunk(layer, seq_len) {
        const ret = wasm.wasmgemma4pipeline_forward_range_ffn_up_chunk(this.__wbg_ptr, layer, seq_len);
        if (ret[1]) {
            throw takeFromExternrefTable0(ret[0]);
        }
    }
    /**
     * Borrow-construct from a host backend. Same `&WasmHostBackend` rule
     * as WasmLlamaPipeline (taking by value triggers wasm-bindgen
     * `__destroy_into_raw` on the JS handle and invalidates subsequent
     * `set_tensor` calls — see WasmLlamaPipeline::from_backend).
     * `kv_base_layer` / `kv_num_layers`: fluid-memory KV alloc (#31). On
     * gemma4-31b a single-layer trisplit worker passing `(17, 1)` cuts
     * the cache from ~503 MiB to ~8 MiB and is the difference between
     * fitting and OOMing on the 128 MiB Worker isolate cap.
     * @param {WasmHostBackend} backend
     * @param {string | null | undefined} config_json
     * @param {number} payload_offset
     * @param {number | null} [kv_base_layer]
     * @param {number | null} [kv_num_layers]
     * @returns {WasmGemma4Pipeline}
     */
    static from_backend(backend, config_json, payload_offset, kv_base_layer, kv_num_layers) {
        _assertClass(backend, WasmHostBackend);
        var ptr0 = isLikeNone(config_json) ? 0 : passStringToWasm0(config_json, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        var len0 = WASM_VECTOR_LEN;
        const ret = wasm.wasmgemma4pipeline_from_backend(backend.__wbg_ptr, ptr0, len0, payload_offset, isLikeNone(kv_base_layer) ? 0x100000001 : (kv_base_layer) >>> 0, isLikeNone(kv_num_layers) ? 0x100000001 : (kv_num_layers) >>> 0);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return WasmGemma4Pipeline.__wrap(ret[0]);
    }
    /**
     * @param {number} seq_len
     * @returns {Float32Array}
     */
    get_batch_hb(seq_len) {
        const ret = wasm.wasmgemma4pipeline_get_batch_hb(this.__wbg_ptr, seq_len);
        var v1 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * @returns {number}
     */
    get_batch_hb_in_ptr() {
        const ret = wasm.wasmgemma4pipeline_get_batch_hb_in_ptr(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * Wave-23 Bug B (deeper fix, Path B): caller-provided output buffer.
     * JS allocates a Float32Array of length `seq_len * intermediate_size`
     * and passes it in; wasm fills it via `Float32Array::copy_from(&slice)`,
     * which uses `set` semantics writing INTO a JS-owned ArrayBuffer.
     *
     * Why this is safer than the `Vec<f32>`-returning `get_batch_hb`:
     * the `Vec<f32>` return path runs (1) Rust `to_vec()`, (2)
     * wasm-bindgen `__wbindgen_malloc` for an export staging buffer in
     * WASM linear memory, (3) JS-side
     * `getFloat32ArrayMemory0().subarray(ptr/4, ptr/4 + len)`, (4)
     * `.slice()` to copy out, (5) `__wbindgen_free`. Steps (2)+(3) are
     * the failure surface: if `__wbindgen_malloc` triggers `memory.grow`
     * mid-call, the cached Float32Array view JS uses for `subarray`
     * becomes invalid even with the wave-22 buffer-identity shim — the
     * shim runs at view-fetch time, but the subsequent `subarray` op
     * can still race a grow inside the same call frame. With this API,
     * JS owns the destination buffer and `Float32Array::copy_from`
     * writes into it without any export-side malloc or subarray call.
     *
     * Returns `Err` if `out.length() < seq_len * intermediate_size`.
     * Path A (boot-time pre-grow in `from_knot_with_kv_range`) is the
     * primary defense; this Path B export is the belt-and-suspenders
     * fallback if Path A doesn't fully eliminate Bug B post-deploy.
     * Existing `get_batch_hb` callers are unaffected — JS migrates
     * only when ready.
     * @param {number} seq_len
     * @param {Float32Array} out
     */
    get_batch_hb_into(seq_len, out) {
        const ret = wasm.wasmgemma4pipeline_get_batch_hb_into(this.__wbg_ptr, seq_len, out);
        if (ret[1]) {
            throw takeFromExternrefTable0(ret[0]);
        }
    }
    /**
     * @returns {number}
     */
    get_batch_x_in_ptr() {
        const ret = wasm.wasmgemma4pipeline_get_batch_x_in_ptr(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * @param {number} seq_len
     * @returns {Float32Array}
     */
    get_batch_xb(seq_len) {
        const ret = wasm.wasmgemma4pipeline_get_batch_xb(this.__wbg_ptr, seq_len);
        var v1 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * @returns {number}
     */
    get_batch_xb_in_ptr() {
        const ret = wasm.wasmgemma4pipeline_get_batch_xb_in_ptr(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * @param {number} seq_len
     * @returns {Float32Array}
     */
    get_residual(seq_len) {
        const ret = wasm.wasmgemma4pipeline_get_residual(this.__wbg_ptr, seq_len);
        var v1 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * True when `layer` is a global-attention layer (Gemma3/4 sliding-
     * window dispatch). Surfaced for diagnostic dumps in
     * `gemma4-pipeline-smoke` and the worker boot.
     * @param {number} layer
     * @returns {boolean}
     */
    is_global_layer(layer) {
        const ret = wasm.wasmgemma4pipeline_is_global_layer(this.__wbg_ptr, layer);
        return ret !== 0;
    }
    /**
     * @param {Float32Array} input
     * @param {number} seq_len
     * @returns {Float32Array}
     */
    lm_head_only(input, seq_len) {
        const ptr0 = passArrayF32ToWasm0(input, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.wasmgemma4pipeline_lm_head_only(this.__wbg_ptr, ptr0, len0, seq_len);
        if (ret[3]) {
            throw takeFromExternrefTable0(ret[2]);
        }
        var v2 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v2;
    }
    /**
     * @param {Float32Array} input
     */
    set_batch_hb(input) {
        const ptr0 = passArrayF32ToWasm0(input, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        wasm.wasmgemma4pipeline_set_batch_hb(this.__wbg_ptr, ptr0, len0);
    }
    /**
     * @param {Float32Array} input
     */
    set_batch_x(input) {
        const ptr0 = passArrayF32ToWasm0(input, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        wasm.wasmgemma4pipeline_set_batch_x(this.__wbg_ptr, ptr0, len0);
    }
    /**
     * Fused split-c setter (wave-22 Fix A). Takes the concatenated
     * `[batch_x | batch_hb]` buffer in a single wasm-bindgen call —
     * halving the JS↔wasm boundary cost vs back-to-back
     * `set_batch_x` + `set_batch_hb`. See SPLIT_C_BOTTLENECK_INVESTIGATION
     * for the +6.0% TPS prediction on gemma4-31b trisplit.
     * @param {Float32Array} combined
     * @param {number} hidden_dim
     * @param {number} seq_len
     */
    set_batch_x_and_hb(combined, hidden_dim, seq_len) {
        const ptr0 = passArrayF32ToWasm0(combined, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        wasm.wasmgemma4pipeline_set_batch_x_and_hb(this.__wbg_ptr, ptr0, len0, hidden_dim, seq_len);
    }
    /**
     * @param {Float32Array} input
     */
    set_batch_xb(input) {
        const ptr0 = passArrayF32ToWasm0(input, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        wasm.wasmgemma4pipeline_set_batch_xb(this.__wbg_ptr, ptr0, len0);
    }
    /**
     * Wave-23 Bug B Path E: combined split-b call. Replaces the
     * `set_batch_xb` + `forward_range_ffn_gate_chunk` +
     * `forward_range_ffn_up_chunk` + `get_batch_hb*` four-call sequence
     * with ONE wasm-bindgen export.
     *
     * The four-call sequence was the residual failure mode after Paths
     * A-D: each call is a separate JS frame, and `memory.grow` could
     * fire between two of them, leaving JS-side typed-array views or
     * caller-provided Float32Array buffers stale by the time the next
     * call ran. Path E collapses everything into one synchronous wasm
     * execution. The return-`Vec<f32>` allocation that wasm-bindgen
     * stages via `__wbindgen_malloc` is the LAST allocation in the
     * frame, so the JS-side `getArrayF32FromWasm0` read sees a stable
     * `wasm.memory.buffer` (no further grow can race it).
     *
     * `batch_xb` is the input residual (`seq_len × hidden_dim` floats).
     * Returns `batch_hb` (`seq_len × intermediate_size` floats).
     * @param {Float32Array} batch_xb
     * @param {number} layer
     * @param {number} seq_len
     * @returns {Float32Array}
     */
    split_b_chunk(batch_xb, layer, seq_len) {
        const ptr0 = passArrayF32ToWasm0(batch_xb, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.wasmgemma4pipeline_split_b_chunk(this.__wbg_ptr, ptr0, len0, layer, seq_len);
        if (ret[3]) {
            throw takeFromExternrefTable0(ret[2]);
        }
        var v2 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v2;
    }
}
if (Symbol.dispose) WasmGemma4Pipeline.prototype[Symbol.dispose] = WasmGemma4Pipeline.prototype.free;
exports.WasmGemma4Pipeline = WasmGemma4Pipeline;

/**
 * JS-facing host-populated weight backend.
 *
 * Constructed from a stringified `tensors` map (the same shape produced by
 * `parse_tensors_json`). The host populates per-tensor bytes via
 * `set_tensor` before invoking the pipeline, and frees them via `evict`.
 */
class WasmHostBackend {
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        WasmHostBackendFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_wasmhostbackend_free(ptr, 0);
    }
    /**
     * Drop tensor bytes from the cache. Called by the host shim after a
     * layer chunk has been consumed by the forward pass.
     * @param {string[]} names
     */
    evict(names) {
        const ptr0 = passArrayJsValueToWasm0(names, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        wasm.wasmhostbackend_evict(this.__wbg_ptr, ptr0, len0);
    }
    /**
     * `metadata_json` is stringified JSON of either:
     *   - the full knot metadata object (with a top-level `tensors` key), or
     *   - the `tensors` map directly (object of `name → { offset, length, type }`).
     * Both shapes are accepted; the constructor probes for `tensors` first.
     * @param {string} metadata_json
     */
    constructor(metadata_json) {
        const ptr0 = passStringToWasm0(metadata_json, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.wasmhostbackend_new(ptr0, len0);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        this.__wbg_ptr = ret[0] >>> 0;
        WasmHostBackendFinalization.register(this, this.__wbg_ptr, this);
        return this;
    }
    /**
     * Insert tensor bytes into the cache. Called once per tensor per layer
     * chunk by the JS host before re-entering the pipeline.
     * @param {string} name
     * @param {Uint8Array} bytes
     */
    set_tensor(name, bytes) {
        const ptr0 = passStringToWasm0(name, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ptr1 = passArray8ToWasm0(bytes, wasm.__wbindgen_malloc);
        const len1 = WASM_VECTOR_LEN;
        wasm.wasmhostbackend_set_tensor(this.__wbg_ptr, ptr0, len0, ptr1, len1);
    }
    /**
     * Return `{ offset, length, tensor_type }` for the given tensor name as
     * a JSON string, or `None` if not in the metadata. JSON-string return
     * (not `JsValue`) sidesteps a wasm-bindgen externref-shim crash where
     * every call to `tensor_info` fired `__rt::throw_null` from inside the
     * generated wrapper — even though the Rust body never produced a null
     * JsValue. Repro stack:
     * `WasmHostBackend.tensor_info → wasmhostbackend_tensor_info externref
     *  shim → __rt::throw_null → __wbindgen_throw`. Returning a primitive
     * `Option<String>` avoids externref entirely and the JS host parses the
     * JSON. The encode + parse cost (~30 bytes once per layer phase) is
     * invisible compared to the per-layer R2 fetch.
     * @param {string} name
     * @returns {string | undefined}
     */
    tensor_info(name) {
        const ptr0 = passStringToWasm0(name, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.wasmhostbackend_tensor_info(this.__wbg_ptr, ptr0, len0);
        let v2;
        if (ret[0] !== 0) {
            v2 = getStringFromWasm0(ret[0], ret[1]).slice();
            wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        }
        return v2;
    }
}
if (Symbol.dispose) WasmHostBackend.prototype[Symbol.dispose] = WasmHostBackend.prototype.free;
exports.WasmHostBackend = WasmHostBackend;

/**
 * JS-facing pipeline. Wraps the cross-target `NativeLlamaPipeline`. The JS
 * host owns one of these per active session; the underlying Rust struct
 * holds the resolved tensor refs, KV cache, and reusable scratch buffers.
 */
class WasmLlamaPipeline {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(WasmLlamaPipeline.prototype);
        obj.__wbg_ptr = ptr;
        WasmLlamaPipelineFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        WasmLlamaPipelineFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_wasmllamapipeline_free(ptr, 0);
    }
    /**
     * @param {bigint} prefix_hash
     * @param {number} prefix_len
     * @param {number} layer_lo
     * @param {number} layer_hi
     * @param {Float32Array} tail_residual
     * @param {Float32Array} kv_slab
     */
    amplituhedron_capture(prefix_hash, prefix_len, layer_lo, layer_hi, tail_residual, kv_slab) {
        const ptr0 = passArrayF32ToWasm0(tail_residual, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        const ptr1 = passArrayF32ToWasm0(kv_slab, wasm.__wbindgen_malloc);
        const len1 = WASM_VECTOR_LEN;
        wasm.wasmllamapipeline_amplituhedron_capture(this.__wbg_ptr, prefix_hash, prefix_len, layer_lo, layer_hi, ptr0, len0, ptr1, len1);
    }
    amplituhedron_clear() {
        wasm.wasmllamapipeline_amplituhedron_clear(this.__wbg_ptr);
    }
    /**
     * @param {boolean} on
     */
    amplituhedron_enable(on) {
        wasm.wasmllamapipeline_amplituhedron_enable(this.__wbg_ptr, on);
    }
    /**
     * On hit, returns `[tail_len, tail..., kv_slab...]`; on miss, returns
     * an empty vector. The first element is exact for current hidden sizes.
     * @param {bigint} prefix_hash
     * @param {number} prefix_len
     * @param {number} layer_lo
     * @param {number} layer_hi
     * @returns {Float32Array}
     */
    amplituhedron_replay(prefix_hash, prefix_len, layer_lo, layer_hi) {
        const ret = wasm.wasmllamapipeline_amplituhedron_replay(this.__wbg_ptr, prefix_hash, prefix_len, layer_lo, layer_hi);
        var v1 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * @param {number} n
     */
    amplituhedron_set_max_volumes(n) {
        wasm.wasmllamapipeline_amplituhedron_set_max_volumes(this.__wbg_ptr, n);
    }
    /**
     * @returns {Uint32Array}
     */
    amplituhedron_stats() {
        const ret = wasm.wasmllamapipeline_amplituhedron_stats(this.__wbg_ptr);
        var v1 = getArrayU32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Embedding-only forward. Entry stations call this for the first chunk
     * (or each batched chunk) to produce the initial residual stream.
     * @param {Uint32Array} tokens
     * @param {number} start_pos
     * @returns {Float32Array}
     */
    embed(tokens, start_pos) {
        const ptr0 = passArray32ToWasm0(tokens, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.wasmllamapipeline_embed(this.__wbg_ptr, ptr0, len0, start_pos);
        if (ret[3]) {
            throw takeFromExternrefTable0(ret[2]);
        }
        var v2 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v2;
    }
    /**
     * Vocabulary-sharded embedding. Returns a residual stream where tokens
     * in `[vocab_start, vocab_end)` are dequantized normally and tokens
     * outside that range produce zero vectors. The coordinator adds all
     * partial results from N cobordism shards (Frobenius μ: A ⊗ A → A).
     * @param {Uint32Array} tokens
     * @param {number} start_pos
     * @param {number} vocab_start
     * @param {number} vocab_end
     * @returns {Float32Array}
     */
    embed_range(tokens, start_pos, vocab_start, vocab_end) {
        const ptr0 = passArray32ToWasm0(tokens, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.wasmllamapipeline_embed_range(this.__wbg_ptr, ptr0, len0, start_pos, vocab_start, vocab_end);
        if (ret[3]) {
            throw takeFromExternrefTable0(ret[2]);
        }
        var v2 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v2;
    }
    /**
     * Run a contiguous range of transformer layers over an incoming residual.
     * This is the load-bearing primitive for log-rolling: each station calls
     * `forward_range` over its pinned layer range and hands off the result to
     * the next station.
     *
     * Causal masking is applied internally for `sequence_length > 1`. KV
     * cache is updated for positions `[start_pos, start_pos + sequence_length)`.
     * @param {Float32Array} residual
     * @param {number} start_layer
     * @param {number} end_layer
     * @param {number} start_pos
     * @param {number} sequence_length
     * @returns {Float32Array}
     */
    forward_range(residual, start_layer, end_layer, start_pos, sequence_length) {
        const ptr0 = passArrayF32ToWasm0(residual, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.wasmllamapipeline_forward_range(this.__wbg_ptr, ptr0, len0, start_layer, end_layer, start_pos, sequence_length);
        if (ret[3]) {
            throw takeFromExternrefTable0(ret[2]);
        }
        var v2 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v2;
    }
    /**
     * Split phase 1: Attention. Uses `[attn_norm, q, k, v, o]`.
     * @param {Float32Array} residual
     * @param {number} layer
     * @param {number} start_pos
     * @param {number} sequence_length
     * @param {boolean} is_first_layer
     */
    forward_range_attn_chunk(residual, layer, start_pos, sequence_length, is_first_layer) {
        const ptr0 = passArrayF32ToWasm0(residual, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.wasmllamapipeline_forward_range_attn_chunk(this.__wbg_ptr, ptr0, len0, layer, start_pos, sequence_length, is_first_layer);
        if (ret[1]) {
            throw takeFromExternrefTable0(ret[0]);
        }
    }
    /**
     * Split phase 3: FFN Down. Uses `[ffn_down]`.
     * Returns the residual vector if `is_last_layer` is true, otherwise an empty array.
     * @param {number} layer
     * @param {number} sequence_length
     * @param {boolean} is_last_layer
     * @returns {Float32Array}
     */
    forward_range_ffn_down_chunk(layer, sequence_length, is_last_layer) {
        const ret = wasm.wasmllamapipeline_forward_range_ffn_down_chunk(this.__wbg_ptr, layer, sequence_length, is_last_layer);
        if (ret[3]) {
            throw takeFromExternrefTable0(ret[2]);
        }
        var v1 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Split FFN phase 2 into 2a (gate) and 2b (up + swiglu). The
     * worker is expected to call them sequentially, evicting the
     * gate matrix between the two — see the rationale comment on
     * `NativeLlamaPipeline::forward_range_ffn_gate_chunk`.
     * @param {number} layer
     * @param {number} sequence_length
     */
    forward_range_ffn_gate_chunk(layer, sequence_length) {
        const ret = wasm.wasmllamapipeline_forward_range_ffn_gate_chunk(this.__wbg_ptr, layer, sequence_length);
        if (ret[1]) {
            throw takeFromExternrefTable0(ret[0]);
        }
    }
    /**
     * Combined gate + up + swiglu path. Retained for callers that
     * don't fluid-memory the FFN halves.
     * @param {number} layer
     * @param {number} sequence_length
     */
    forward_range_ffn_gate_up_chunk(layer, sequence_length) {
        const ret = wasm.wasmllamapipeline_forward_range_ffn_gate_up_chunk(this.__wbg_ptr, layer, sequence_length);
        if (ret[1]) {
            throw takeFromExternrefTable0(ret[0]);
        }
    }
    /**
     * @param {number} layer
     * @param {number} sequence_length
     */
    forward_range_ffn_up_chunk(layer, sequence_length) {
        const ret = wasm.wasmllamapipeline_forward_range_ffn_up_chunk(this.__wbg_ptr, layer, sequence_length);
        if (ret[1]) {
            throw takeFromExternrefTable0(ret[0]);
        }
    }
    /**
     * Borrow-construct from a backend. We take `&WasmHostBackend` (not by
     * value) so the JS handle stays alive for subsequent
     * `set_tensor` / `evict` / `tensor_info` calls. Fluid-memory boot
     * loads tensors per layer phase, then forwards, then evicts — that
     * pattern requires the backend handle to remain valid past pipeline
     * construction.
     *
     * Previous shape `pub fn from_backend(backend: WasmHostBackend, ...)`
     * caused a `__rt::throw_null` ("null pointer passed to rust") panic
     * on every set_tensor that ran after pipeline construction; the
     * wasm-bindgen-generated `_assertClass + __destroy_into_raw` chain
     * zeros `__wbg_ptr` when the wrapper is taken by value. Switching
     * to `&WasmHostBackend` keeps the wrapper alive on the JS side.
     *
     * `config_json` is either a stringified `KnotConfig` or `null`. Passing
     * `null` falls back to TinyLlama defaults (matches the legacy 1B-knot
     * path; see `TransformerConfig::default` in `model.rs`).
     * `kv_base_layer` / `kv_num_layers`: optional fluid-memory KV alloc.
     * Pass `(layer_index_we_own, 1)` from a single-layer trisplit worker
     * to shrink the KV cache from full-model to one-layer (~503 MiB →
     * ~8 MiB on gemma4-31b global heads). `None` defaults preserve legacy
     * whole-model alloc — back-compat for callers that pre-date #31.
     * @param {WasmHostBackend} backend
     * @param {string | null | undefined} config_json
     * @param {number} payload_offset
     * @param {boolean} draft_mode
     * @param {number} resonance_threshold
     * @param {number | null} [kv_base_layer]
     * @param {number | null} [kv_num_layers]
     * @returns {WasmLlamaPipeline}
     */
    static from_backend(backend, config_json, payload_offset, draft_mode, resonance_threshold, kv_base_layer, kv_num_layers) {
        _assertClass(backend, WasmHostBackend);
        var ptr0 = isLikeNone(config_json) ? 0 : passStringToWasm0(config_json, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        var len0 = WASM_VECTOR_LEN;
        const ret = wasm.wasmllamapipeline_from_backend(backend.__wbg_ptr, ptr0, len0, payload_offset, draft_mode, resonance_threshold, isLikeNone(kv_base_layer) ? 0x100000001 : (kv_base_layer) >>> 0, isLikeNone(kv_num_layers) ? 0x100000001 : (kv_num_layers) >>> 0);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return WasmLlamaPipeline.__wrap(ret[0]);
    }
    /**
     * Run the autoregressive generation loop. Returns the newly generated
     * tokens (excludes the prompt). Same semantics as the native NAPI path.
     * @param {Uint32Array} prompt_tokens
     * @param {number} max_new_tokens
     * @returns {Uint32Array}
     */
    generate(prompt_tokens, max_new_tokens) {
        const ptr0 = passArray32ToWasm0(prompt_tokens, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.wasmllamapipeline_generate(this.__wbg_ptr, ptr0, len0, max_new_tokens);
        if (ret[3]) {
            throw takeFromExternrefTable0(ret[2]);
        }
        var v2 = getArrayU32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v2;
    }
    /**
     * @param {number} seq_len
     * @returns {Float32Array}
     */
    get_batch_hb(seq_len) {
        const ret = wasm.wasmllamapipeline_get_batch_hb(this.__wbg_ptr, seq_len);
        var v1 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * @returns {number}
     */
    get_batch_hb_in_ptr() {
        const ret = wasm.wasmllamapipeline_get_batch_hb_in_ptr(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * @returns {number}
     */
    get_batch_x_in_ptr() {
        const ret = wasm.wasmllamapipeline_get_batch_x_in_ptr(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * @param {number} seq_len
     * @returns {Float32Array}
     */
    get_batch_xb(seq_len) {
        const ret = wasm.wasmllamapipeline_get_batch_xb(this.__wbg_ptr, seq_len);
        var v1 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * @returns {number}
     */
    get_batch_xb_in_ptr() {
        const ret = wasm.wasmllamapipeline_get_batch_xb_in_ptr(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * @param {number} layer_lo
     * @param {number} layer_hi
     * @param {number} prefix_len
     * @returns {Float32Array}
     */
    get_kv_slab(layer_lo, layer_hi, prefix_len) {
        const ret = wasm.wasmllamapipeline_get_kv_slab(this.__wbg_ptr, layer_lo, layer_hi, prefix_len);
        var v1 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Current decode position (number of tokens fed so far).
     * @returns {number}
     */
    get_pos() {
        const ret = wasm.wasmllamapipeline_get_pos(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * Read the first `seq_len * hidden_dim` floats out of the pipeline's
     * `batch_x` buffer. Used by the liquid-mesh `/forward-one` path:
     * the worker calls `forward_range_ffn_down_chunk(layer, seq_len, false)`
     * to leave the post-residual-add output in `batch_x`, then calls this
     * to extract the residual. Returning `Vec<f32>` from the kernel
     * directly (with `is_last_layer=true`) was throwing
     * "Invalid array buffer length" on the JS side after the kernel
     * completed — likely a wasm-bindgen multivalue-return path issue
     * triggered by some interaction between this kernel's `vec!` and
     * the post-grow memory state. Splitting the access avoids it.
     * @param {number} seq_len
     * @returns {Float32Array}
     */
    get_residual(seq_len) {
        const ret = wasm.wasmllamapipeline_get_residual(this.__wbg_ptr, seq_len);
        var v1 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Apply final norm + lm_head to a residual stream and return logits for
     * the LAST token in the chunk (the next-token-prediction row). Exit
     * stations call this.
     * @param {Float32Array} residual
     * @param {number} sequence_length
     * @returns {Float32Array}
     */
    lm_head(residual, sequence_length) {
        const ptr0 = passArrayF32ToWasm0(residual, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.wasmllamapipeline_lm_head(this.__wbg_ptr, ptr0, len0, sequence_length);
        if (ret[3]) {
            throw takeFromExternrefTable0(ret[2]);
        }
        var v2 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v2;
    }
    /**
     * Fluid-memory vocabulary-sharded lm_head. The host pre-loads only
     * the rows of `output_weight` for `[vocab_start, vocab_end)` under
     * the synthetic name `lm_head_partial_weight`. Peak resident weight
     * bytes is one shard's slice (~55 MiB Q6_K for an 8-way qwen-7b
     * shard) — fits inside the 128 MiB Worker cap, where the full
     * 447 MiB tensor cannot.
     *
     * See `NativeLlamaPipeline::lm_head_partial` for the host-side flow.
     * @param {Float32Array} residual
     * @param {number} sequence_length
     * @param {number} vocab_start
     * @param {number} vocab_end
     * @returns {Float32Array}
     */
    lm_head_partial(residual, sequence_length, vocab_start, vocab_end) {
        const ptr0 = passArrayF32ToWasm0(residual, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.wasmllamapipeline_lm_head_partial(this.__wbg_ptr, ptr0, len0, sequence_length, vocab_start, vocab_end);
        if (ret[3]) {
            throw takeFromExternrefTable0(ret[2]);
        }
        var v2 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v2;
    }
    /**
     * Vocabulary-sharded lm_head. Computes partial logits over
     * `[vocab_start, vocab_end)` only, returning a vector of length
     * `(vocab_end - vocab_start)`. The coordinator concatenates partials
     * from all cobordism shards (comultiplication Δ: A → A ⊗ A).
     * @param {Float32Array} residual
     * @param {number} sequence_length
     * @param {number} vocab_start
     * @param {number} vocab_end
     * @returns {Float32Array}
     */
    lm_head_range(residual, sequence_length, vocab_start, vocab_end) {
        const ptr0 = passArrayF32ToWasm0(residual, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.wasmllamapipeline_lm_head_range(this.__wbg_ptr, ptr0, len0, sequence_length, vocab_start, vocab_end);
        if (ret[3]) {
            throw takeFromExternrefTable0(ret[2]);
        }
        var v2 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v2;
    }
    memo_clear() {
        wasm.wasmllamapipeline_memo_clear(this.__wbg_ptr);
    }
    /**
     * Per-(layer, op) hit-rate breakdown. Flat layout, four u32s per
     * entry: `[layer, op, hits, misses, layer, op, hits, misses, ...]`.
     * `op` is the u8 op-id from the `OP_*` constants in `matvec_memo.rs`,
     * widened to u32 for the JS Uint32Array boundary. Iteration order
     * follows the underlying HashMap and is therefore not stable
     * across calls; the bench should sort by (layer, op) before
     * rendering. Length is always a multiple of 4. Empty when no
     * probes have run.
     * @returns {Uint32Array}
     */
    memo_detailed_stats() {
        const ret = wasm.wasmllamapipeline_memo_detailed_stats(this.__wbg_ptr);
        var v1 = getArrayU32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Pisot drift histogram (Death #5 — Connes-Kreimer/Pisot residue).
     * Returns `[luminary, total, drift_min_q12, drift_mean_q12, drift_max_q12]`.
     * `drift_*_q12 = (drift * 4096.0) as u32` — fixed-point, JS divides by 4096.
     * `luminary_threshold_q12` likewise: caller passes `(threshold * 4096) as u32`.
     * Honest scope: this is the f32 surrogate flagged at
     * `matvec_memo::pisot_drift`. The Lean witness lives on integer Lucas
     * pairs (`PisotMitosisManifold.mitosis_restores_invariant`); this
     * accessor exposes the surrogate distribution, not a Hopf-algebraic
     * renormalization. Use it to detect when the working set drifts off
     * the φ-attractor — that is the operational meaning of "halting
     * residue" the FORMAL_LEDGER claims for RMSNorm but RMSNorm itself
     * does not directly compute.
     * @param {number} luminary_threshold_q12
     * @returns {Uint32Array}
     */
    memo_drift_summary(luminary_threshold_q12) {
        const ret = wasm.wasmllamapipeline_memo_drift_summary(this.__wbg_ptr, luminary_threshold_q12);
        var v1 = getArrayU32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * @param {boolean} on
     */
    memo_enable(on) {
        wasm.wasmllamapipeline_memo_enable(this.__wbg_ptr, on);
    }
    /**
     * Direct Pisot-drift probe — compute the surrogate on an arbitrary
     * f32 vector WITHOUT inserting into the cache. Useful for validating
     * the surrogate scale on healthy inputs without needing a working
     * kernel chain. Returns drift × 4096 in q12 fixed-point; JS divides
     * by 4096 to recover f32. Returns u32::MAX for non-finite drifts.
     * @param {Float32Array} x
     * @returns {number}
     */
    memo_probe_drift(x) {
        const ptr0 = passArrayF32ToWasm0(x, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.wasmllamapipeline_memo_probe_drift(this.__wbg_ptr, ptr0, len0);
        return ret >>> 0;
    }
    /**
     * @param {number} n
     */
    memo_set_max_entries(n) {
        wasm.wasmllamapipeline_memo_set_max_entries(this.__wbg_ptr, n);
    }
    /**
     * @param {number} tau_sq
     */
    memo_set_tau_squared(tau_sq) {
        wasm.wasmllamapipeline_memo_set_tau_squared(this.__wbg_ptr, tau_sq);
    }
    /**
     * Aggregate stats packed as a 3-element Vec<u32>:
     * `[hits, misses, evictions]`.
     * @returns {Uint32Array}
     */
    memo_stats() {
        const ret = wasm.wasmllamapipeline_memo_stats(this.__wbg_ptr);
        var v1 = getArrayU32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * @param {Float32Array} input
     */
    set_batch_hb(input) {
        const ptr0 = passArrayF32ToWasm0(input, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        wasm.wasmllamapipeline_set_batch_hb(this.__wbg_ptr, ptr0, len0);
    }
    /**
     * @param {Float32Array} input
     */
    set_batch_x(input) {
        const ptr0 = passArrayF32ToWasm0(input, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        wasm.wasmllamapipeline_set_batch_x(this.__wbg_ptr, ptr0, len0);
    }
    /**
     * Fused split-c setter (wave-22 Fix A). Takes the concatenated
     * `[batch_x | batch_hb]` buffer in a single wasm-bindgen call —
     * halving the JS↔wasm boundary cost vs back-to-back
     * `set_batch_x` + `set_batch_hb`. See SPLIT_C_BOTTLENECK_INVESTIGATION
     * for the +6.0% TPS prediction on gemma4-31b trisplit.
     * @param {Float32Array} combined
     * @param {number} hidden_dim
     * @param {number} seq_len
     */
    set_batch_x_and_hb(combined, hidden_dim, seq_len) {
        const ptr0 = passArrayF32ToWasm0(combined, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        wasm.wasmllamapipeline_set_batch_x_and_hb(this.__wbg_ptr, ptr0, len0, hidden_dim, seq_len);
    }
    /**
     * @param {Float32Array} input
     */
    set_batch_xb(input) {
        const ptr0 = passArrayF32ToWasm0(input, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        wasm.wasmllamapipeline_set_batch_xb(this.__wbg_ptr, ptr0, len0);
    }
    /**
     * @param {number} layer_lo
     * @param {number} layer_hi
     * @param {number} prefix_len
     * @param {Float32Array} slab
     */
    set_kv_slab(layer_lo, layer_hi, prefix_len, slab) {
        const ptr0 = passArrayF32ToWasm0(slab, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        wasm.wasmllamapipeline_set_kv_slab(this.__wbg_ptr, layer_lo, layer_hi, prefix_len, ptr0, len0);
    }
    /**
     * Wave-23 Bug B Path E: combined split-b call for the LLaMA arch
     * (qwen2 / qwen-coder-7b / llama / mistral). Mirrors
     * `WasmGemma4Pipeline::split_b_chunk`. Replaces the four-call
     * `set_batch_xb` + `forward_range_ffn_gate_chunk` +
     * `forward_range_ffn_up_chunk` + `get_batch_hb*` JS-side sequence
     * with ONE wasm-bindgen export.
     *
     * See `docs/bug-b-path-e-result.md` for the full investigation —
     * the four-call sequence was the residual failure mode after
     * Paths A-D: each call is a separate JS frame, and `memory.grow`
     * could fire between two of them, leaving JS-side typed-array
     * views or caller-provided Float32Array buffers stale by the
     * time the next call ran. Path E collapses everything into one
     * synchronous wasm execution. The return-`Vec<f32>` allocation
     * that wasm-bindgen stages via `__wbindgen_malloc` is the LAST
     * allocation in the frame, so the JS-side `getArrayF32FromWasm0`
     * read sees a stable `wasm.memory.buffer` (no further grow can
     * race it).
     *
     * `batch_xb` is the input residual (`seq_len × hidden_dim` floats).
     * Returns `batch_hb` (`seq_len × intermediate_size` floats) — the
     * SwiGLU-fused gate*up output ready to feed `/split-c`.
     * @param {Float32Array} batch_xb
     * @param {number} layer
     * @param {number} seq_len
     * @returns {Float32Array}
     */
    split_b_chunk(batch_xb, layer, seq_len) {
        const ptr0 = passArrayF32ToWasm0(batch_xb, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.wasmllamapipeline_split_b_chunk(this.__wbg_ptr, ptr0, len0, layer, seq_len);
        if (ret[3]) {
            throw takeFromExternrefTable0(ret[2]);
        }
        var v2 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v2;
    }
}
if (Symbol.dispose) WasmLlamaPipeline.prototype[Symbol.dispose] = WasmLlamaPipeline.prototype.free;
exports.WasmLlamaPipeline = WasmLlamaPipeline;

/**
 * Confidence + provenance JSON from [`build_mesh_depth_field`] (default config, ONNX off).
 */
class WasmMeshDepthStyleResult {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(WasmMeshDepthStyleResult.prototype);
        obj.__wbg_ptr = ptr;
        WasmMeshDepthStyleResultFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        WasmMeshDepthStyleResultFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_wasmmeshdepthstyleresult_free(ptr, 0);
    }
    /**
     * @returns {Float32Array}
     */
    get confidence() {
        const ret = wasm.wasmmeshdepthstyleresult_confidence(this.__wbg_ptr);
        var v1 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * @returns {string}
     */
    get provenance_json() {
        let deferred1_0;
        let deferred1_1;
        try {
            const ret = wasm.wasmmeshdepthstyleresult_provenance_json(this.__wbg_ptr);
            deferred1_0 = ret[0];
            deferred1_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        } finally {
            wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
        }
    }
}
if (Symbol.dispose) WasmMeshDepthStyleResult.prototype[Symbol.dispose] = WasmMeshDepthStyleResult.prototype.free;
exports.WasmMeshDepthStyleResult = WasmMeshDepthStyleResult;

/**
 * Result of [`wasm_mesh_residual_deblur_rgb_f32`]: reconstructed RGB,
 * admission gate, and a serde certificate for host-side inspection.
 */
class WasmMeshResidualDeblurResult {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(WasmMeshResidualDeblurResult.prototype);
        obj.__wbg_ptr = ptr;
        WasmMeshResidualDeblurResultFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        WasmMeshResidualDeblurResultFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_wasmmeshresidualdeblurresult_free(ptr, 0);
    }
    /**
     * @returns {Float32Array}
     */
    get admission() {
        const ret = wasm.wasmmeshresidualdeblurresult_admission(this.__wbg_ptr);
        var v1 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * @returns {string}
     */
    get certificate_json() {
        let deferred1_0;
        let deferred1_1;
        try {
            const ret = wasm.wasmmeshresidualdeblurresult_certificate_json(this.__wbg_ptr);
            deferred1_0 = ret[0];
            deferred1_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        } finally {
            wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
        }
    }
    /**
     * @returns {Float32Array}
     */
    get rgb() {
        const ret = wasm.wasmmeshresidualdeblurresult_rgb(this.__wbg_ptr);
        var v1 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
}
if (Symbol.dispose) WasmMeshResidualDeblurResult.prototype[Symbol.dispose] = WasmMeshResidualDeblurResult.prototype.free;
exports.WasmMeshResidualDeblurResult = WasmMeshResidualDeblurResult;

/**
 * Result of [`wasm_motion_deblur_rgb_f32`]: deblurred RGB, a serde certificate,
 * and diagnostic layers flattened into a manifest + concatenated f32 buffer.
 */
class WasmMotionDeblurResult {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(WasmMotionDeblurResult.prototype);
        obj.__wbg_ptr = ptr;
        WasmMotionDeblurResultFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        WasmMotionDeblurResultFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_wasmmotiondeblurresult_free(ptr, 0);
    }
    /**
     * @returns {string}
     */
    get certificate_json() {
        let deferred1_0;
        let deferred1_1;
        try {
            const ret = wasm.wasmmotiondeblurresult_certificate_json(this.__wbg_ptr);
            deferred1_0 = ret[0];
            deferred1_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        } finally {
            wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
        }
    }
    /**
     * All layer planes concatenated in manifest order; each plane is
     * `width*height` f32 in `[0,1]`.
     * @returns {Float32Array}
     */
    get layers_data() {
        const ret = wasm.wasmmotiondeblurresult_layers_data(this.__wbg_ptr);
        var v1 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * JSON array manifest: one entry per layer with
     * `{name,label,width,height,colormap,offset,len}` indexing into `layers_data`
     * (offset/len in f32 elements).
     * @returns {string}
     */
    get layers_json() {
        let deferred1_0;
        let deferred1_1;
        try {
            const ret = wasm.wasmmotiondeblurresult_layers_json(this.__wbg_ptr);
            deferred1_0 = ret[0];
            deferred1_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        } finally {
            wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
        }
    }
    /**
     * @returns {Float32Array}
     */
    get rgb() {
        const ret = wasm.wasmmotiondeblurresult_rgb(this.__wbg_ptr);
        var v1 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
}
if (Symbol.dispose) WasmMotionDeblurResult.prototype[Symbol.dispose] = WasmMotionDeblurResult.prototype.free;
exports.WasmMotionDeblurResult = WasmMotionDeblurResult;

/**
 * JS-facing Pneuma TTS pipeline. Holds an acoustic model + vocoder pair
 * resolved against two host-populated weight backends.
 */
class WasmPneumaTtsPipeline {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(WasmPneumaTtsPipeline.prototype);
        obj.__wbg_ptr = ptr;
        WasmPneumaTtsPipelineFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        WasmPneumaTtsPipelineFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_wasmpneumattspipeline_free(ptr, 0);
    }
    /**
     * Construct from two backends + their metadata JSON blobs.
     *
     * `acoustic_metadata_json` and `vocoder_metadata_json` are the
     * stringified `KNOT` metadata sections (the same JSON the JS host
     * already parses from the knot header). The acoustic blob carries
     * the `TtsConfig` schema and the per-knot codec marker; the vocoder
     * blob carries `VocoderConfig`. Schemas live in
     * `model_pneuma_tts_acoustic.rs` / `model_pneuma_tts_vocoder.rs`.
     * @param {WasmHostBackend} acoustic_backend
     * @param {WasmHostBackend} vocoder_backend
     * @param {string} acoustic_metadata_json
     * @param {string} vocoder_metadata_json
     * @returns {WasmPneumaTtsPipeline}
     */
    static from_backends(acoustic_backend, vocoder_backend, acoustic_metadata_json, vocoder_metadata_json) {
        _assertClass(acoustic_backend, WasmHostBackend);
        _assertClass(vocoder_backend, WasmHostBackend);
        const ptr0 = passStringToWasm0(acoustic_metadata_json, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ptr1 = passStringToWasm0(vocoder_metadata_json, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len1 = WASM_VECTOR_LEN;
        const ret = wasm.wasmpneumattspipeline_from_backends(acoustic_backend.__wbg_ptr, vocoder_backend.__wbg_ptr, ptr0, len0, ptr1, len1);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return WasmPneumaTtsPipeline.__wrap(ret[0]);
    }
    /**
     * Sample rate the vocoder will emit. Matches what the route handler
     * puts in the `X-Sample-Rate` response header. Read once per
     * session; doesn't change after construction.
     * @returns {number}
     */
    sample_rate() {
        const ret = wasm.wasmpneumattspipeline_sample_rate(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * Synthesize PCM from a phoneme-id stream.
     *
     * Same end-to-end shape as the smoke binary
     * (`pneuma-tts-smoke.rs`):
     *
     *   acoustic.forward(phonemes) -> mel
     *   vocoder.forward(mel)       -> pcm
     *
     * Returns the f32 mono PCM samples; sample rate is reported via
     * `sample_rate()`. Errors on empty input / empty mel / empty pcm
     * so the route handler can surface a 4xx/5xx with provenance
     * instead of shipping silence.
     * @param {Uint32Array} phoneme_ids
     * @returns {Float32Array}
     */
    synthesize_from_phonemes(phoneme_ids) {
        const ptr0 = passArray32ToWasm0(phoneme_ids, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.wasmpneumattspipeline_synthesize_from_phonemes(this.__wbg_ptr, ptr0, len0);
        if (ret[3]) {
            throw takeFromExternrefTable0(ret[2]);
        }
        var v2 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v2;
    }
}
if (Symbol.dispose) WasmPneumaTtsPipeline.prototype[Symbol.dispose] = WasmPneumaTtsPipeline.prototype.free;
exports.WasmPneumaTtsPipeline = WasmPneumaTtsPipeline;

/**
 * Result of [`wasm_relight_monster_mesh_masked`]: relit RGB buffer.
 */
class WasmRelightMonsterMeshResult {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(WasmRelightMonsterMeshResult.prototype);
        obj.__wbg_ptr = ptr;
        WasmRelightMonsterMeshResultFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        WasmRelightMonsterMeshResultFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_wasmrelightmonstermeshresult_free(ptr, 0);
    }
    /**
     * @returns {Float32Array}
     */
    get rgb() {
        const ret = wasm.wasmrelightmonstermeshresult_rgb(this.__wbg_ptr);
        var v1 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
}
if (Symbol.dispose) WasmRelightMonsterMeshResult.prototype[Symbol.dispose] = WasmRelightMonsterMeshResult.prototype.free;
exports.WasmRelightMonsterMeshResult = WasmRelightMonsterMeshResult;

/**
 * Result of [`sfm_solve`]: a JSON document describing either the recovered pose
 * or an honest null. Shape:
 *
 * ```json
 * // pose:
 * { "ok": true, "r": [[..],[..],[..]], "t": [x,y,z],
 *   "intrinsics": {"fx":..,"fy":..,"cx":..,"cy":..},
 *   "inliers": 73, "matches": 91, "medianParallaxDeg": 6.2, "confidence": 0.71 }
 * // honest null:
 * { "ok": false, "reason": "...", "confidence": 0.0 }
 * ```
 */
class WasmSfmResult {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(WasmSfmResult.prototype);
        obj.__wbg_ptr = ptr;
        WasmSfmResultFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        WasmSfmResultFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_wasmsfmresult_free(ptr, 0);
    }
    /**
     * @returns {number}
     */
    get confidence() {
        const ret = wasm.wasmsfmresult_confidence(this.__wbg_ptr);
        return ret;
    }
    /**
     * `true` when a trustworthy pose was recovered; `false` = honest null
     * (caller should fall back, e.g. to the turntable carve).
     * @returns {boolean}
     */
    get isPose() {
        const ret = wasm.wasmsfmresult_isPose(this.__wbg_ptr);
        return ret !== 0;
    }
    /**
     * @returns {string}
     */
    get json() {
        let deferred1_0;
        let deferred1_1;
        try {
            const ret = wasm.wasmsfmresult_json(this.__wbg_ptr);
            deferred1_0 = ret[0];
            deferred1_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        } finally {
            wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
        }
    }
}
if (Symbol.dispose) WasmSfmResult.prototype[Symbol.dispose] = WasmSfmResult.prototype.free;
exports.WasmSfmResult = WasmSfmResult;

/**
 * Result of Topological Calibration (Manifold Shift or LUT).
 */
class WasmTopologicalCalibrationResult {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(WasmTopologicalCalibrationResult.prototype);
        obj.__wbg_ptr = ptr;
        WasmTopologicalCalibrationResultFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        WasmTopologicalCalibrationResultFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_wasmtopologicalcalibrationresult_free(ptr, 0);
    }
    /**
     * @returns {Float32Array}
     */
    get rgb() {
        const ret = wasm.wasmtopologicalcalibrationresult_rgb(this.__wbg_ptr);
        var v1 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
}
if (Symbol.dispose) WasmTopologicalCalibrationResult.prototype[Symbol.dispose] = WasmTopologicalCalibrationResult.prototype.free;
exports.WasmTopologicalCalibrationResult = WasmTopologicalCalibrationResult;

/**
 * Result of [`wasm_topology_mask_build_rgb_f32`]: soft alpha `[0,1]^wh` plus serde provenance JSON.
 */
class WasmTopologyMaskResult {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(WasmTopologyMaskResult.prototype);
        obj.__wbg_ptr = ptr;
        WasmTopologyMaskResultFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        WasmTopologyMaskResultFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_wasmtopologymaskresult_free(ptr, 0);
    }
    /**
     * @returns {Float32Array}
     */
    get alpha() {
        const ret = wasm.wasmtopologymaskresult_alpha(this.__wbg_ptr);
        var v1 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * @returns {string}
     */
    get provenance_json() {
        let deferred1_0;
        let deferred1_1;
        try {
            const ret = wasm.wasmtopologymaskresult_provenance_json(this.__wbg_ptr);
            deferred1_0 = ret[0];
            deferred1_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        } finally {
            wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
        }
    }
}
if (Symbol.dispose) WasmTopologyMaskResult.prototype[Symbol.dispose] = WasmTopologyMaskResult.prototype.free;
exports.WasmTopologyMaskResult = WasmTopologyMaskResult;

/**
 * JS-callable Whisper STT pipeline.
 *
 * Construction is two-step: the JS host first builds a
 * `WasmHostBackend` (same backend used for the LLM path) and
 * pre-populates the whisper tensor names, then constructs this
 * wrapper. Encoding and decoding share the backend handle —
 * nothing special-cased on the JS side.
 */
class WasmWhisperPipeline {
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        WasmWhisperPipelineFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_wasmwhisperpipeline_free(ptr, 0);
    }
    /**
     * Run one decoder step and return the real next-token logits.
     * @param {number} token
     * @param {number} pos
     * @returns {Float32Array}
     */
    decode_step_logits(token, pos) {
        const ret = wasm.wasmwhisperpipeline_decode_step_logits(this.__wbg_ptr, token, pos);
        var v1 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Run the encoder. `mel` is a `Float32Array` of length
     * `n_mels * n_frames` (typically `80 * 3000 = 240_000`).
     * Returns a `Float32Array` of length
     * `n_audio_ctx * encoder_hidden_dim` (typically
     * `1500 * 384 = 576_000` for whisper-tiny).
     * @param {Float32Array} mel
     * @returns {Float32Array}
     */
    encode(mel) {
        const ptr0 = passArrayF32ToWasm0(mel, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.wasmwhisperpipeline_encode(this.__wbg_ptr, ptr0, len0);
        var v2 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v2;
    }
    /**
     * Greedy autoregressive decode. Returns the generated token
     * ids (BOS excluded). Stops at `eos_token` or after
     * `max_tokens` steps, whichever fires first.
     * @param {Float32Array} encoder_output
     * @param {number} max_tokens
     * @param {number} eos_token
     * @returns {Uint32Array}
     */
    generate(encoder_output, max_tokens, eos_token) {
        const ptr0 = passArrayF32ToWasm0(encoder_output, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.wasmwhisperpipeline_generate(this.__wbg_ptr, ptr0, len0, max_tokens, eos_token);
        var v2 = getArrayU32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v2;
    }
    /**
     * Prime cross-attention K/V from an encoder output and reset the
     * autoregressive self-attention cache for a new utterance.
     * @param {Float32Array} encoder_output
     */
    load_audio_context(encoder_output) {
        const ptr0 = passArrayF32ToWasm0(encoder_output, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        wasm.wasmwhisperpipeline_load_audio_context(this.__wbg_ptr, ptr0, len0);
    }
    memo_clear() {
        wasm.wasmwhisperpipeline_memo_clear(this.__wbg_ptr);
    }
    /**
     * @param {boolean} on
     */
    memo_enable(on) {
        wasm.wasmwhisperpipeline_memo_enable(this.__wbg_ptr, on);
    }
    /**
     * @returns {Uint32Array}
     */
    memo_stats() {
        const ret = wasm.wasmwhisperpipeline_memo_stats(this.__wbg_ptr);
        var v1 = getArrayU32FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Construct from a `WasmHostBackend` and a JSON config string.
     *
     * `backend` is taken by reference (not by value) so the JS
     * handle stays alive for `set_tensor` / `evict` calls that
     * run between the encoder pass and the decode loop. Same
     * rationale as `WasmLlamaPipeline::from_backend` — taking
     * the wrapper by value triggers a `__rt::throw_null` panic
     * on the next `set_tensor` because wasm-bindgen zeros
     * `__wbg_ptr` on consumed wrappers.
     *
     * `config_json` is the stringified `WhisperConfig` (see Agent
     * D). Either `{ "config": {...} }` or the bare config object
     * is accepted.
     * @param {WasmHostBackend} backend
     * @param {string} config_json
     */
    constructor(backend, config_json) {
        _assertClass(backend, WasmHostBackend);
        const ptr0 = passStringToWasm0(config_json, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.wasmwhisperpipeline_new(backend.__wbg_ptr, ptr0, len0);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        this.__wbg_ptr = ret[0] >>> 0;
        WasmWhisperPipelineFinalization.register(this, this.__wbg_ptr, this);
        return this;
    }
    /**
     * Reset only the autoregressive decoder self-attention cache.
     */
    reset_decoder() {
        wasm.wasmwhisperpipeline_reset_decoder(this.__wbg_ptr);
    }
    /**
     * Override the BOS token used by the next `generate` call.
     * Defaults to `50258` (multilingual). Set to `50257` for
     * `whisper-*.en` checkpoints, or any other special-token id
     * for forced-prompt experiments.
     * @param {number} bos
     */
    set_bos_token(bos) {
        wasm.wasmwhisperpipeline_set_bos_token(this.__wbg_ptr, bos);
    }
    /**
     * Override the forced decoder prefix used by the next `generate`
     * call. Defaults to `[50258, 50259, 50359, 50363]`
     * (`<|sot|><|en|><|transcribe|><|notimestamps|>`). Pass an empty
     * array to restore legacy BOS-only seeding, or a custom prefix
     * for other languages / translation / timestamped decoding.
     * @param {Uint32Array} prefix
     */
    set_forced_prefix(prefix) {
        const ptr0 = passArray32ToWasm0(prefix, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        wasm.wasmwhisperpipeline_set_forced_prefix(this.__wbg_ptr, ptr0, len0);
    }
}
if (Symbol.dispose) WasmWhisperPipeline.prototype[Symbol.dispose] = WasmWhisperPipeline.prototype.free;
exports.WasmWhisperPipeline = WasmWhisperPipeline;

/**
 * F-mesh-2 instrumentation (COLTRANE_MESH_FALSIFICATION_LEDGER): install
 * the panic hook at WASM module instantiation, BEFORE any export is
 * called. The previous regime installed it lazily inside
 * `WasmHostBackend::new` and `WasmGemma4Pipeline::from_backend`, which
 * left a window where `parse_tensors_json` / `parse_config_json` (the
 * FIRST exports the worker boot calls — see
 * `apps/distributed-inference-worker/src/index.ts:327-334`) could
 * panic with the runtime emitting a bare `RuntimeError: unreachable`
 * and no source line. With `--target web`, wasm-bindgen invokes
 * `start` synchronously as part of `wasmGlue.default(wasmModule)`,
 * so by the time the JS host calls any export the hook is already
 * in place. `set_once` is idempotent so leaving the existing
 * `install_panic_hook()` calls in `new` / `from_backend` is harmless
 * belt-and-braces — kept so a future caller that bypasses module
 * init (e.g. `WebAssembly.instantiate` without the bindgen wrapper)
 * still gets the hook on first use.
 */
function __dii_init() {
    wasm.__dii_init();
}
exports.__dii_init = __dii_init;

/**
 * @param {any} v1_val
 * @param {any} v2_val
 * @returns {any}
 */
function buley_route(v1_val, v2_val) {
    const ret = wasm.buley_route(v1_val, v2_val);
    if (ret[2]) {
        throw takeFromExternrefTable0(ret[1]);
    }
    return takeFromExternrefTable0(ret[0]);
}
exports.buley_route = buley_route;

/**
 * @param {any} source_val
 * @param {any} attractor_val
 * @returns {any}
 */
function buley_siphon(source_val, attractor_val) {
    const ret = wasm.buley_siphon(source_val, attractor_val);
    if (ret[2]) {
        throw takeFromExternrefTable0(ret[1]);
    }
    return takeFromExternrefTable0(ret[0]);
}
exports.buley_siphon = buley_siphon;

/**
 * 6-char hex for `bw_cache_fp24_frf_v1` final digest.
 * @param {Uint8Array} data
 * @returns {string}
 */
function bw_cache_fp24_frf_hex_v1(data) {
    let deferred2_0;
    let deferred2_1;
    try {
        const ptr0 = passArray8ToWasm0(data, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.bw_cache_fp24_frf_hex_v1(ptr0, len0);
        deferred2_0 = ret[0];
        deferred2_1 = ret[1];
        return getStringFromWasm0(ret[0], ret[1]);
    } finally {
        wasm.__wbindgen_free(deferred2_0, deferred2_1, 1);
    }
}
exports.bw_cache_fp24_frf_hex_v1 = bw_cache_fp24_frf_hex_v1;

/**
 * 3-byte little-endian FRF fingerprint, masked to 24 bits.
 * @param {Uint8Array} data
 * @returns {Uint8Array}
 */
function bw_cache_fp24_frf_v1(data) {
    const ptr0 = passArray8ToWasm0(data, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.bw_cache_fp24_frf_v1(ptr0, len0);
    var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v2;
}
exports.bw_cache_fp24_frf_v1 = bw_cache_fp24_frf_v1;

/**
 * 6-char lowercase hex fingerprint (low 24 bits of `bw_cache_fp64_hex_v1`).
 * @param {Uint8Array} data
 * @returns {string}
 */
function bw_cache_fp24_hex_v1(data) {
    let deferred2_0;
    let deferred2_1;
    try {
        const ptr0 = passArray8ToWasm0(data, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.bw_cache_fp24_hex_v1(ptr0, len0);
        deferred2_0 = ret[0];
        deferred2_1 = ret[1];
        return getStringFromWasm0(ret[0], ret[1]);
    } finally {
        wasm.__wbindgen_free(deferred2_0, deferred2_1, 1);
    }
}
exports.bw_cache_fp24_hex_v1 = bw_cache_fp24_hex_v1;

/**
 * 3-byte (24-bit) little-endian cache fingerprint — the byte-clean floor below
 * `bw_cache_fp64_v1`. Half the wire of fp48, a quarter of fp64. UNSOUND
 * standalone (~2^12 birthday); ship only as a context-keyed front channel with
 * verify-on-hit (TS `contextKeyedFp24` / `verifyFp24OnHit`).
 * @param {Uint8Array} data
 * @returns {Uint8Array}
 */
function bw_cache_fp24_v1(data) {
    const ptr0 = passArray8ToWasm0(data, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.bw_cache_fp24_v1(ptr0, len0);
    var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v2;
}
exports.bw_cache_fp24_v1 = bw_cache_fp24_v1;

/**
 * 16-char hex for `bw_cache_fp64_frf_v1` final digest.
 * @param {Uint8Array} data
 * @returns {string}
 */
function bw_cache_fp64_frf_hex_v1(data) {
    let deferred2_0;
    let deferred2_1;
    try {
        const ptr0 = passArray8ToWasm0(data, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.bw_cache_fp64_frf_hex_v1(ptr0, len0);
        deferred2_0 = ret[0];
        deferred2_1 = ret[1];
        return getStringFromWasm0(ret[0], ret[1]);
    } finally {
        wasm.__wbindgen_free(deferred2_0, deferred2_1, 1);
    }
}
exports.bw_cache_fp64_frf_hex_v1 = bw_cache_fp64_frf_hex_v1;

/**
 * @param {Uint8Array} data
 * @returns {Uint8Array}
 */
function bw_cache_fp64_frf_v1(data) {
    const ptr0 = passArray8ToWasm0(data, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.bw_cache_fp64_frf_v1(ptr0, len0);
    var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v2;
}
exports.bw_cache_fp64_frf_v1 = bw_cache_fp64_frf_v1;

/**
 * 16-char lowercase hex fingerprint (same algorithm as `bw_cache_fp64_v1`).
 * @param {Uint8Array} data
 * @returns {string}
 */
function bw_cache_fp64_hex_v1(data) {
    let deferred2_0;
    let deferred2_1;
    try {
        const ptr0 = passArray8ToWasm0(data, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.bw_cache_fp64_hex_v1(ptr0, len0);
        deferred2_0 = ret[0];
        deferred2_1 = ret[1];
        return getStringFromWasm0(ret[0], ret[1]);
    } finally {
        wasm.__wbindgen_free(deferred2_0, deferred2_1, 1);
    }
}
exports.bw_cache_fp64_hex_v1 = bw_cache_fp64_hex_v1;

/**
 * 8-byte little-endian cache fingerprint (XXH64). Smaller KV keys than SHA-256 hex.
 * @param {Uint8Array} data
 * @returns {Uint8Array}
 */
function bw_cache_fp64_v1(data) {
    const ptr0 = passArray8ToWasm0(data, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.bw_cache_fp64_v1(ptr0, len0);
    var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v2;
}
exports.bw_cache_fp64_v1 = bw_cache_fp64_v1;

/**
 * Compute compression ratio: text bytes vs packed bytes.
 * Per zero_overhead theorem: packed ≤ text (always).
 * @param {number} text_length
 * @returns {number}
 */
function compression_ratio(text_length) {
    const ret = wasm.compression_ratio(text_length);
    return ret;
}
exports.compression_ratio = compression_ratio;

/**
 * Decode a FlowFrame flags byte into (structure_type, bit_width_nibble).
 * @param {number} flags
 * @returns {Uint8Array}
 */
function decode_flags(flags) {
    const ret = wasm.decode_flags(flags);
    var v1 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v1;
}
exports.decode_flags = decode_flags;

/**
 * Decode a FlowFrame header, returning
 * [stream_id_hi, stream_id_lo, pos_b3, pos_b2, pos_b1, pos_b0,
 *  structure_type, bit_width_nibble, len_b2, len_b1, len_b0].
 * @param {Uint8Array} frame
 * @returns {Uint8Array}
 */
function decode_flowframe_header(frame) {
    const ptr0 = passArray8ToWasm0(frame, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.decode_flowframe_header(ptr0, len0);
    var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v2;
}
exports.decode_flowframe_header = decode_flowframe_header;

/**
 * Decode a FlowFrame stream_id into (dimension, chromosome).
 * @param {number} stream_id
 * @returns {Uint8Array}
 */
function decode_helix_stream_id(stream_id) {
    const ret = wasm.decode_helix_stream_id(stream_id);
    var v1 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v1;
}
exports.decode_helix_stream_id = decode_helix_stream_id;

/**
 * Generate dilated sampling positions for a genomic range.
 *
 * Given a start position, length, and dilation strategy, returns
 * the positions at which this dimension should be sampled.
 * @param {number} start
 * @param {number} length
 * @param {number} stride
 * @returns {Uint32Array}
 */
function dilated_positions(start, length, stride) {
    const ret = wasm.dilated_positions(start, length, stride);
    var v1 = getArrayU32FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
    return v1;
}
exports.dilated_positions = dilated_positions;

/**
 * Encode a FlowFrame flags byte from structure type and bit-width.
 *
 * flags = (structure_type << 4) | bit_width_nibble
 *
 * Structure types (upper nibble):
 *   0x1_ = PROCESS (default stream data)
 *   0x2_ = FORK    (stream split)
 *   0x4_ = FOLD    (stream merge)
 *   0x8_ = VENT    (stream termination)
 * @param {number} structure_type
 * @param {number} bit_width_nibble
 * @returns {number}
 */
function encode_flags(structure_type, bit_width_nibble) {
    const ret = wasm.encode_flags(structure_type, bit_width_nibble);
    return ret;
}
exports.encode_flags = encode_flags;

/**
 * Encode packed DNA bases into an Aeon FlowFrame.
 *
 * FlowFrame header (10 bytes):
 *   [stream_id:2][sequence:4][flags:1][length:3]
 *
 * stream_id = chromosome number (1-25)
 * sequence  = genomic position / window_size
 * flags     = structure type
 * payload   = packed bases for this window
 * @param {number} chromosome
 * @param {number} position
 * @param {number} flags
 * @param {Uint8Array} payload
 * @returns {Uint8Array}
 */
function encode_flowframe(chromosome, position, flags, payload) {
    const ptr0 = passArray8ToWasm0(payload, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.encode_flowframe(chromosome, position, flags, ptr0, len0);
    var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v2;
}
exports.encode_flowframe = encode_flowframe;

/**
 * Encode a dimension index and chromosome into a FlowFrame stream_id.
 *
 * Layout: `(dimension << 8) | chromosome`
 * - dimension: 0-53 (upper byte)
 * - chromosome: 1-25 typical (lower byte)
 * @param {number} dimension
 * @param {number} chromosome
 * @returns {number}
 */
function encode_helix_stream_id(dimension, chromosome) {
    const ret = wasm.encode_helix_stream_id(dimension, chromosome);
    return ret;
}
exports.encode_helix_stream_id = encode_helix_stream_id;

/**
 * Encode a typed FlowFrame with explicit bit-width in the flags byte.
 *
 * This is the general-purpose version of encode_flowframe that embeds
 * the encoding width so decoders know how to interpret the payload.
 * @param {number} stream_id
 * @param {number} position
 * @param {number} structure_type
 * @param {number} bit_width_nibble
 * @param {Uint8Array} payload
 * @returns {Uint8Array}
 */
function encode_typed_flowframe(stream_id, position, structure_type, bit_width_nibble, payload) {
    const ptr0 = passArray8ToWasm0(payload, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.encode_typed_flowframe(stream_id, position, structure_type, bit_width_nibble, ptr0, len0);
    var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v2;
}
exports.encode_typed_flowframe = encode_typed_flowframe;

/**
 * Convert a full FASTA sequence into a stream of FlowFrames.
 * One frame per window_size bases.
 * @param {Uint8Array} sequence
 * @param {number} chromosome
 * @param {number} start_position
 * @param {number} window_size
 * @returns {Uint8Array}
 */
function fasta_to_flowframes(sequence, chromosome, start_position, window_size) {
    const ptr0 = passArray8ToWasm0(sequence, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.fasta_to_flowframes(ptr0, len0, chromosome, start_position, window_size);
    var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v2;
}
exports.fasta_to_flowframes = fasta_to_flowframes;

/**
 * Compute the 3D helical coordinates (x, y, z) for a genomic position.
 *
 * x = R * cos(phase)
 * y = R * sin(phase)
 * z = position * rise_per_bp
 *
 * where R = 10 angstroms (radius of B-form DNA double helix).
 * Returns [x, y, z] in angstroms.
 * @param {number} position
 * @returns {Float64Array}
 */
function helix_coordinates(position) {
    const ret = wasm.helix_coordinates(position);
    var v1 = getArrayF64FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 8, 8);
    return v1;
}
exports.helix_coordinates = helix_coordinates;

/**
 * Compute the helical phase angle (radians) for a genomic position.
 *
 * The B-form DNA helix completes one full turn every 10.5 bp.
 * Phase = (position mod 10.5) * (2π / 10.5)
 *
 * This maps each base to its rotational position around the helix axis,
 * enabling rotationally-aware sampling and analysis.
 * @param {number} position
 * @returns {number}
 */
function helix_phase(position) {
    const ret = wasm.helix_phase(position);
    return ret;
}
exports.helix_phase = helix_phase;

/**
 * Q4_K dispatch — symmetry with `mat_vec_q6k_dispatch` for shards whose
 * `output_weight` (or fallback `token_embd_weight`) is Q4_K rather than
 * Q6_K. Goes through the wave-22 wasm32 SIMD `mat_vec_q4k_row`.
 * @param {Float32Array} x
 * @param {Uint8Array} w
 * @param {number} n
 * @param {number} d
 * @returns {Float32Array}
 */
function mat_vec_q4k_dispatch(x, w, n, d) {
    const ptr0 = passArrayF32ToWasm0(x, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passArray8ToWasm0(w, wasm.__wbindgen_malloc);
    const len1 = WASM_VECTOR_LEN;
    const ret = wasm.mat_vec_q4k_dispatch(ptr0, len0, ptr1, len1, n, d);
    if (ret[3]) {
        throw takeFromExternrefTable0(ret[2]);
    }
    var v3 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
    return v3;
}
exports.mat_vec_q4k_dispatch = mat_vec_q4k_dispatch;

/**
 * Q6_K dispatch — calls `crate::math::mat_vec_q6k` over the supplied
 * residual + quantised weight slice. The wasm32 build picks up the
 * SIMD branch in `mat_vec_q6k_row` automatically.
 * @param {Float32Array} x
 * @param {Uint8Array} w
 * @param {number} n
 * @param {number} d
 * @returns {Float32Array}
 */
function mat_vec_q6k_dispatch(x, w, n, d) {
    const ptr0 = passArrayF32ToWasm0(x, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passArray8ToWasm0(w, wasm.__wbindgen_malloc);
    const len1 = WASM_VECTOR_LEN;
    const ret = wasm.mat_vec_q6k_dispatch(ptr0, len0, ptr1, len1, n, d);
    if (ret[3]) {
        throw takeFromExternrefTable0(ret[2]);
    }
    var v3 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
    return v3;
}
exports.mat_vec_q6k_dispatch = mat_vec_q6k_dispatch;

/**
 * Count mutations between two packed sequences via XOR.
 * Per xor_detects_mutations: XOR = 0 iff sequences are identical.
 * Returns the number of differing bases.
 * @param {Uint8Array} seq_a
 * @param {Uint8Array} seq_b
 * @param {number} length
 * @returns {number}
 */
function mutation_count(seq_a, seq_b, length) {
    const ptr0 = passArray8ToWasm0(seq_a, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passArray8ToWasm0(seq_b, wasm.__wbindgen_malloc);
    const len1 = WASM_VECTOR_LEN;
    const ret = wasm.mutation_count(ptr0, len0, ptr1, len1, length);
    return ret >>> 0;
}
exports.mutation_count = mutation_count;

/**
 * Pack values into 4-bit encoding. 2 values per byte, MSB first.
 * Values are clamped to 0-15. For 16-level quantized data.
 * @param {Uint8Array} values
 * @returns {Uint8Array}
 */
function pack_4bit(values) {
    const ptr0 = passArray8ToWasm0(values, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.pack_4bit(ptr0, len0);
    var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v2;
}
exports.pack_4bit = pack_4bit;

/**
 * Pack annotation values (0-3 each, 2 bits) into bytes, same as base packing.
 * This reuses the 2-bit packing scheme for all dimension channels:
 * value 0 = low/absent, 1 = low-medium, 2 = medium-high, 3 = high/present.
 * @param {Uint8Array} values
 * @returns {Uint8Array}
 */
function pack_annotation(values) {
    const ptr0 = passArray8ToWasm0(values, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.pack_annotation(ptr0, len0);
    var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v2;
}
exports.pack_annotation = pack_annotation;

/**
 * Pack a DNA sequence into 2-bit binary. 4 bases per byte.
 * Per two_bit_four_per_byte theorem: 4 * 2 = 8.
 * @param {Uint8Array} sequence
 * @returns {Uint8Array}
 */
function pack_bases(sequence) {
    const ptr0 = passArray8ToWasm0(sequence, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.pack_bases(ptr0, len0);
    var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v2;
}
exports.pack_bases = pack_bases;

/**
 * Pack f64 values into bytes (big-endian IEEE 754).
 * @param {Float64Array} values
 * @returns {Uint8Array}
 */
function pack_f64(values) {
    const ptr0 = passArrayF64ToWasm0(values, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.pack_f64(ptr0, len0);
    var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v2;
}
exports.pack_f64 = pack_f64;

/**
 * Pack u16 values into bytes (big-endian).
 * @param {Uint16Array} values
 * @returns {Uint8Array}
 */
function pack_u16(values) {
    const ptr0 = passArray16ToWasm0(values, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.pack_u16(ptr0, len0);
    var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v2;
}
exports.pack_u16 = pack_u16;

/**
 * Pack u32 values into bytes (big-endian).
 * @param {Uint32Array} values
 * @returns {Uint8Array}
 */
function pack_u32(values) {
    const ptr0 = passArray32ToWasm0(values, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.pack_u32(ptr0, len0);
    var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v2;
}
exports.pack_u32 = pack_u32;

/**
 * Parse a knot metadata JSON string and return the optional `config` block
 * as a stringified `KnotConfig`, or `null` if absent.
 * @param {string} metadata_json
 * @returns {string | undefined}
 */
function parse_config_json(metadata_json) {
    const ptr0 = passStringToWasm0(metadata_json, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.parse_config_json(ptr0, len0);
    if (ret[3]) {
        throw takeFromExternrefTable0(ret[2]);
    }
    let v2;
    if (ret[0] !== 0) {
        v2 = getStringFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    }
    return v2;
}
exports.parse_config_json = parse_config_json;

/**
 * Parse a FASTA record: extract header and pack the sequence.
 * Returns (header, packed_bases, sequence_length).
 * @param {Uint8Array} input
 * @returns {any}
 */
function parse_fasta_record(input) {
    const ptr0 = passArray8ToWasm0(input, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.parse_fasta_record(ptr0, len0);
    return ret;
}
exports.parse_fasta_record = parse_fasta_record;

/**
 * Parse a knot metadata JSON string and return the `tensors` map as a
 * stringified JSON object: `{ name: { offset, length, tensor_type }, ... }`.
 * Lets the JS host keep the parsed map without re-implementing the field
 * renames (`type` → `tensor_type`) on the JS side.
 * @param {string} metadata_json
 * @returns {string}
 */
function parse_tensors_json(metadata_json) {
    let deferred3_0;
    let deferred3_1;
    try {
        const ptr0 = passStringToWasm0(metadata_json, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.parse_tensors_json(ptr0, len0);
        var ptr2 = ret[0];
        var len2 = ret[1];
        if (ret[3]) {
            ptr2 = 0; len2 = 0;
            throw takeFromExternrefTable0(ret[2]);
        }
        deferred3_0 = ptr2;
        deferred3_1 = len2;
        return getStringFromWasm0(ptr2, len2);
    } finally {
        wasm.__wbindgen_free(deferred3_0, deferred3_1, 1);
    }
}
exports.parse_tensors_json = parse_tensors_json;

/**
 * Perfect-phase-lock predicate: returns `true` iff `r ∈ {17, 34, 51}`.
 * Mirrors the Lean `is_perfect_phase_lock` decidable. Use this to gate
 * `perfect_phase_lock_witness_score_r` (which returns `0` for both
 * "not in set" and "score=0", a degenerate but legal witness reading).
 * @param {number} r
 * @returns {boolean}
 */
function perfect_phase_lock_check_r(r) {
    const ret = wasm.perfect_phase_lock_check_r(r);
    return ret !== 0;
}
exports.perfect_phase_lock_check_r = perfect_phase_lock_check_r;

/**
 * Score of `constructive_interference(a, b)` for two `BuleyUnit`s
 * supplied as flat `(rim, void, rigidity)` triples. Pointwise
 * saturating add of all three fields, then sum. Mirrors the Lean
 * `constructive_interference` definition.
 * @param {number} a_rim
 * @param {number} a_void
 * @param {number} a_rigidity
 * @param {number} b_rim
 * @param {number} b_void
 * @param {number} b_rigidity
 * @returns {number}
 */
function perfect_phase_lock_constructive(a_rim, a_void, a_rigidity, b_rim, b_void, b_rigidity) {
    const ret = wasm.perfect_phase_lock_constructive(a_rim, a_void, a_rigidity, b_rim, b_void, b_rigidity);
    return ret >>> 0;
}
exports.perfect_phase_lock_constructive = perfect_phase_lock_constructive;

/**
 * Score of `destructive_interference(a, b)` for two `BuleyUnit`s
 * supplied as flat `(rim, void, rigidity)` triples. Pointwise
 * saturating subtract of all three fields, then sum. At `a == b`
 * (and in particular at the witness `⟨r, 0, 0⟩` against itself for
 * any perfect `r`) this returns `0` — the death-of-interference
 * collapse from Lean `death_of_interference`.
 * @param {number} a_rim
 * @param {number} a_void
 * @param {number} a_rigidity
 * @param {number} b_rim
 * @param {number} b_void
 * @param {number} b_rigidity
 * @returns {number}
 */
function perfect_phase_lock_destructive(a_rim, a_void, a_rigidity, b_rim, b_void, b_rigidity) {
    const ret = wasm.perfect_phase_lock_destructive(a_rim, a_void, a_rigidity, b_rim, b_void, b_rigidity);
    return ret >>> 0;
}
exports.perfect_phase_lock_destructive = perfect_phase_lock_destructive;

/**
 * Score of the Lean death-of-interference witness at density `r`.
 * Returns `witness.score()` (which equals `r`) when `r` is a perfect
 * density, else `0`. Pair with `perfect_phase_lock_check_r` to
 * distinguish the "no witness" case from a legitimate zero score.
 * @param {number} r
 * @returns {number}
 */
function perfect_phase_lock_witness_score_r(r) {
    const ret = wasm.perfect_phase_lock_witness_score_r(r);
    return ret >>> 0;
}
exports.perfect_phase_lock_witness_score_r = perfect_phase_lock_witness_score_r;

/**
 * Evaluate the scattering amplitude `⟨pluckerVector(plane), dual⟩`.
 * `dual` is zip-truncated against the Plücker vector (List.zip semantics),
 * matching the Lean `scatteringAmplitude` definition. Returns `0` on
 * validation failure (the empty inner product) — pair with
 * `scattering_kplane_validate` if you need to distinguish "real zero"
 * from "validation failure".
 * @param {number} k
 * @param {number} d
 * @param {BigInt64Array} rows_flat
 * @param {BigInt64Array} dual
 * @returns {bigint}
 */
function scattering_amplitude_eval(k, d, rows_flat, dual) {
    const ptr0 = passArray64ToWasm0(rows_flat, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passArray64ToWasm0(dual, wasm.__wbindgen_malloc);
    const len1 = WASM_VECTOR_LEN;
    const ret = wasm.scattering_amplitude_eval(k, d, ptr0, len0, ptr1, len1);
    return ret;
}
exports.scattering_amplitude_eval = scattering_amplitude_eval;

/**
 * Decidable positivity check for Gr⁺(k, d) membership. Returns `true`
 * iff every Plücker coordinate is strictly `> 0`. Returns `false` on
 * validation failure (matches the Lean `isPositiveBool` boundary
 * semantics: a malformed plane is not positive).
 * @param {number} k
 * @param {number} d
 * @param {BigInt64Array} rows_flat
 * @returns {boolean}
 */
function scattering_is_positive(k, d, rows_flat) {
    const ptr0 = passArray64ToWasm0(rows_flat, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.scattering_is_positive(k, d, ptr0, len0);
    return ret !== 0;
}
exports.scattering_is_positive = scattering_is_positive;

/**
 * Validate a flattened k-plane: returns `true` iff `rows_flat.len() == k * d`
 * AND `KPlane::new(k, d, ...)` succeeds. Mirrors the implicit precondition
 * of every other `scattering_*` call in this module.
 * @param {number} k
 * @param {number} d
 * @param {BigInt64Array} rows_flat
 * @returns {boolean}
 */
function scattering_kplane_validate(k, d, rows_flat) {
    const ptr0 = passArray64ToWasm0(rows_flat, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.scattering_kplane_validate(k, d, ptr0, len0);
    return ret !== 0;
}
exports.scattering_kplane_validate = scattering_kplane_validate;

/**
 * Compute the Plücker vector of the supplied k-plane. Returns one entry
 * per ordered k-subset of `[0..d)`, in `k_subsets` order (reverse-lex).
 * Returns an empty vector if the plane fails validation — callers should
 * either call `scattering_kplane_validate` first or check `is_empty()`.
 * @param {number} k
 * @param {number} d
 * @param {BigInt64Array} rows_flat
 * @returns {BigInt64Array}
 */
function scattering_plucker_vector(k, d, rows_flat) {
    const ptr0 = passArray64ToWasm0(rows_flat, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.scattering_plucker_vector(k, d, ptr0, len0);
    var v2 = getArrayI64FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 8, 8);
    return v2;
}
exports.scattering_plucker_vector = scattering_plucker_vector;

/**
 * Vertex count of Gr⁺(k, d), i.e. `(k_subsets k d).len()` = `C(d, k)`.
 * Pure combinatorial helper — no plane data needed. Mirrors the Lean
 * `vertexCount` definition; the Lean coverage sweep oracle pins the
 * values `(2,4)→6, (2,6)→15, (2,8)→28, (3,6)→20, (3,9)→84, (3,12)→220`.
 * @param {number} k
 * @param {number} d
 * @returns {number}
 */
function scattering_vertex_count(k, d) {
    const ret = wasm.scattering_vertex_count(k, d);
    return ret >>> 0;
}
exports.scattering_vertex_count = scattering_vertex_count;

/**
 * Search for a DNA pattern in packed binary using word-parallel matching.
 * Compares 32 bases per u64 word via XOR.
 *
 * Per word_parallel_speedup: 32x faster than character-by-character.
 * Per xor_detects_mutations: XOR produces nonzero exactly at mismatches.
 *
 * Returns byte offsets where the pattern matches.
 * @param {Uint8Array} haystack
 * @param {number} haystack_len
 * @param {Uint8Array} needle
 * @param {number} needle_len
 * @returns {Uint32Array}
 */
function search_packed(haystack, haystack_len, needle, needle_len) {
    const ptr0 = passArray8ToWasm0(haystack, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passArray8ToWasm0(needle, wasm.__wbindgen_malloc);
    const len1 = WASM_VECTOR_LEN;
    const ret = wasm.search_packed(ptr0, len0, haystack_len, ptr1, len1, needle_len);
    var v3 = getArrayU32FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
    return v3;
}
exports.search_packed = search_packed;

/**
 * Search for a DNA pattern directly inside a Syzygy-folded structure.
 *
 * Logarithmic performance: Because Syzygy evaluates DNA topologically, it skips
 * over interior data natively. If a structure is 1,000,000 repetitions long,
 * the algorithm logically resolves its interior in O(1) math without iterating
 * characters or allocating expansion memory.
 * @param {Uint8Array} folded
 * @param {Uint8Array} needle
 * @returns {Uint32Array}
 */
function search_syzygy(folded, needle) {
    const ptr0 = passArray8ToWasm0(folded, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passArray8ToWasm0(needle, wasm.__wbindgen_malloc);
    const len1 = WASM_VECTOR_LEN;
    const ret = wasm.search_syzygy(ptr0, len0, ptr1, len1);
    var v3 = getArrayU32FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
    return v3;
}
exports.search_syzygy = search_syzygy;

/**
 * Solve the relative pose of camera2 vs camera1 from two **grayscale** frames
 * (`u8`, row-major, length `width*height` each). Returns a [`WasmSfmResult`]
 * with a JSON pose or an honest null. Never fabricates a pose.
 *
 * `fast_threshold` / `min_parallax_deg` of 0 select the [`SfmConfig::default`]
 * values (handy for the JS caller that doesn't want to tune).
 * @param {Uint8Array} gray1
 * @param {Uint8Array} gray2
 * @param {number} width
 * @param {number} height
 * @param {number} fast_threshold
 * @param {number} min_parallax_deg
 * @returns {WasmSfmResult}
 */
function sfmSolve(gray1, gray2, width, height, fast_threshold, min_parallax_deg) {
    const ptr0 = passArray8ToWasm0(gray1, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passArray8ToWasm0(gray2, wasm.__wbindgen_malloc);
    const len1 = WASM_VECTOR_LEN;
    const ret = wasm.sfmSolve(ptr0, len0, ptr1, len1, width, height, fast_threshold, min_parallax_deg);
    if (ret[2]) {
        throw takeFromExternrefTable0(ret[1]);
    }
    return WasmSfmResult.__wrap(ret[0]);
}
exports.sfmSolve = sfmSolve;

/**
 * Execute a Syzygy Fold (Pass 59) on a raw DNA sequence.
 *
 * Converts a 1D incompressible string into a topological invariant manifold
 * by dilating into 2-bit crossings and executing a structural RLE collapse
 * (e.g. Poly-A -> O(1) complexity).
 * @param {Uint8Array} sequence
 * @returns {Uint8Array}
 */
function syzygy_fold(sequence) {
    const ptr0 = passArray8ToWasm0(sequence, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.syzygy_fold(ptr0, len0);
    var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v2;
}
exports.syzygy_fold = syzygy_fold;

/**
 * Execute a Syzygy Fold on an arbitrary binary sequence (General Purpose Archiver)
 * Drops the DNA-specific 2-bit packing to compress standard desktop byte streams.
 * @param {Uint8Array} sequence
 * @returns {Uint8Array}
 */
function syzygy_fold_binary(sequence) {
    const ptr0 = passArray8ToWasm0(sequence, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.syzygy_fold_binary(ptr0, len0);
    var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v2;
}
exports.syzygy_fold_binary = syzygy_fold_binary;

/**
 * Fold arbitrary binary data into a 54-dimension FlowFrame stream.
 *
 * Each byte is assigned to dimension (byte_index % 54). Bytes within each
 * dimension are windowed into FlowFrames of up to 1024 bytes each. The
 * position field encodes the original byte offset so the unfold can
 * reconstruct exact ordering.
 * @param {Uint8Array} sequence
 * @returns {Uint8Array}
 */
function syzygy_fold_structural(sequence) {
    const ptr0 = passArray8ToWasm0(sequence, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.syzygy_fold_structural(ptr0, len0);
    var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v2;
}
exports.syzygy_fold_structural = syzygy_fold_structural;

/**
 * Smush: Topological Compression followed by AEAD Encryption.
 * Perfect security plus topological efficiency in one pass.
 * @param {Uint8Array} sequence
 * @param {string} key_phrase
 * @returns {Uint8Array}
 */
function syzygy_smush(sequence, key_phrase) {
    const ptr0 = passArray8ToWasm0(sequence, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passStringToWasm0(key_phrase, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
    const len1 = WASM_VECTOR_LEN;
    const ret = wasm.syzygy_smush(ptr0, len0, ptr1, len1);
    var v3 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v3;
}
exports.syzygy_smush = syzygy_smush;

/**
 * Unfold a Syzygy binary buffer back into 1D raw bases.
 * @param {Uint8Array} folded
 * @param {number} original_length
 * @returns {Uint8Array}
 */
function syzygy_unfold(folded, original_length) {
    const ptr0 = passArray8ToWasm0(folded, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.syzygy_unfold(ptr0, len0, original_length);
    var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v2;
}
exports.syzygy_unfold = syzygy_unfold;

/**
 * Unfold a generalized Syzygy binary buffer back into arbitrary byte octets.
 * @param {Uint8Array} folded
 * @param {number} original_length
 * @returns {Uint8Array}
 */
function syzygy_unfold_binary(folded, original_length) {
    const ptr0 = passArray8ToWasm0(folded, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.syzygy_unfold_binary(ptr0, len0, original_length);
    var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v2;
}
exports.syzygy_unfold_binary = syzygy_unfold_binary;

/**
 * Unfold a 54-dimension FlowFrame stream back into raw bytes.
 *
 * Parses the BWFF header, walks FlowFrames, extracts per-dimension payloads,
 * and reassembles bytes in their original order using the round-robin pattern.
 * @param {Uint8Array} stream
 * @param {number} original_length
 * @returns {Uint8Array}
 */
function syzygy_unfold_structural(stream, original_length) {
    const ptr0 = passArray8ToWasm0(stream, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.syzygy_unfold_structural(ptr0, len0, original_length);
    var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v2;
}
exports.syzygy_unfold_structural = syzygy_unfold_structural;

/**
 * Unsmush: Securely decrypt and unfold back to raw octets.
 * @param {Uint8Array} smushed
 * @param {string} key_phrase
 * @param {number} original_length
 * @returns {Uint8Array}
 */
function syzygy_unsmush(smushed, key_phrase, original_length) {
    const ptr0 = passArray8ToWasm0(smushed, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passStringToWasm0(key_phrase, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
    const len1 = WASM_VECTOR_LEN;
    const ret = wasm.syzygy_unsmush(ptr0, len0, ptr1, len1, original_length);
    var v3 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v3;
}
exports.syzygy_unsmush = syzygy_unsmush;

/**
 * Unpack 4-bit values back to a vector.
 * @param {Uint8Array} packed
 * @param {number} length
 * @returns {Uint8Array}
 */
function unpack_4bit(packed, length) {
    const ptr0 = passArray8ToWasm0(packed, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.unpack_4bit(ptr0, len0, length);
    var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v2;
}
exports.unpack_4bit = unpack_4bit;

/**
 * Unpack 2-bit annotation values back to a vector.
 * @param {Uint8Array} packed
 * @param {number} length
 * @returns {Uint8Array}
 */
function unpack_annotation(packed, length) {
    const ptr0 = passArray8ToWasm0(packed, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.unpack_annotation(ptr0, len0, length);
    var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v2;
}
exports.unpack_annotation = unpack_annotation;

/**
 * Unpack 2-bit binary back to DNA sequence.
 * Per pack_unpack theorem: round-trip is identity.
 * @param {Uint8Array} packed
 * @param {number} length
 * @returns {Uint8Array}
 */
function unpack_bases(packed, length) {
    const ptr0 = passArray8ToWasm0(packed, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.unpack_bases(ptr0, len0, length);
    var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v2;
}
exports.unpack_bases = unpack_bases;

/**
 * Unpack bytes into f64 values (big-endian IEEE 754).
 * @param {Uint8Array} packed
 * @param {number} count
 * @returns {Float64Array}
 */
function unpack_f64(packed, count) {
    const ptr0 = passArray8ToWasm0(packed, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.unpack_f64(ptr0, len0, count);
    var v2 = getArrayF64FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 8, 8);
    return v2;
}
exports.unpack_f64 = unpack_f64;

/**
 * Unpack bytes into u16 values (big-endian).
 * @param {Uint8Array} packed
 * @param {number} count
 * @returns {Uint16Array}
 */
function unpack_u16(packed, count) {
    const ptr0 = passArray8ToWasm0(packed, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.unpack_u16(ptr0, len0, count);
    var v2 = getArrayU16FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 2, 2);
    return v2;
}
exports.unpack_u16 = unpack_u16;

/**
 * Unpack bytes into u32 values (big-endian).
 * @param {Uint8Array} packed
 * @param {number} count
 * @returns {Uint32Array}
 */
function unpack_u32(packed, count) {
    const ptr0 = passArray8ToWasm0(packed, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.unpack_u32(ptr0, len0, count);
    var v2 = getArrayU32FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
    return v2;
}
exports.unpack_u32 = unpack_u32;

/**
 * Blend: `0` = SourceOver, `1` = Multiply (see [`BlendMode`]).
 * @param {Float32Array} bg
 * @param {Float32Array} fg
 * @param {Float32Array} alpha
 * @param {number} width
 * @param {number} height
 * @param {number} blend_mode
 * @returns {Float32Array}
 */
function wasmCompositeMaskedRgb(bg, fg, alpha, width, height, blend_mode) {
    const ptr0 = passArrayF32ToWasm0(bg, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passArrayF32ToWasm0(fg, wasm.__wbindgen_malloc);
    const len1 = WASM_VECTOR_LEN;
    const ptr2 = passArrayF32ToWasm0(alpha, wasm.__wbindgen_malloc);
    const len2 = WASM_VECTOR_LEN;
    const ret = wasm.wasmCompositeMaskedRgb(ptr0, len0, ptr1, len1, ptr2, len2, width, height, blend_mode);
    if (ret[3]) {
        throw takeFromExternrefTable0(ret[2]);
    }
    var v4 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
    return v4;
}
exports.wasmCompositeMaskedRgb = wasmCompositeMaskedRgb;

/**
 * Estimate the motion PSF `[angle_deg, length_px]` from row-major linear RGB `[0,1]`.
 * @param {Float32Array} rgb
 * @param {number} width
 * @param {number} height
 * @returns {Float32Array}
 */
function wasmEstimateMotionPsf(rgb, width, height) {
    const ptr0 = passArrayF32ToWasm0(rgb, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.wasmEstimateMotionPsf(ptr0, len0, width, height);
    var v2 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
    return v2;
}
exports.wasmEstimateMotionPsf = wasmEstimateMotionPsf;

/**
 * **Manifold Calibration**: High-dimensional White/Black Balance via WASM.
 * @param {Float32Array} rgb
 * @param {Float32Array} topology_alpha
 * @param {number} width
 * @param {number} height
 * @param {number} r_shift
 * @param {number} g_shift
 * @param {number} b_shift
 * @param {number} pedestal_persistence
 * @returns {WasmTopologicalCalibrationResult}
 */
function wasmManifoldCalibrationApply(rgb, topology_alpha, width, height, r_shift, g_shift, b_shift, pedestal_persistence) {
    const ptr0 = passArrayF32ToWasm0(rgb, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passArrayF32ToWasm0(topology_alpha, wasm.__wbindgen_malloc);
    const len1 = WASM_VECTOR_LEN;
    const ret = wasm.wasmManifoldCalibrationApply(ptr0, len0, ptr1, len1, width, height, r_shift, g_shift, b_shift, pedestal_persistence);
    if (ret[2]) {
        throw takeFromExternrefTable0(ret[1]);
    }
    return WasmTopologicalCalibrationResult.__wrap(ret[0]);
}
exports.wasmManifoldCalibrationApply = wasmManifoldCalibrationApply;

/**
 * Mask multiply: `dst = src * alpha` per RGB channel (`alpha` per pixel).
 * @param {Float32Array} src
 * @param {Float32Array} alpha
 * @param {number} width
 * @param {number} height
 * @returns {Float32Array}
 */
function wasmMaskMultiplyRgb(src, alpha, width, height) {
    const ptr0 = passArrayF32ToWasm0(src, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passArrayF32ToWasm0(alpha, wasm.__wbindgen_malloc);
    const len1 = WASM_VECTOR_LEN;
    const ret = wasm.wasmMaskMultiplyRgb(ptr0, len0, ptr1, len1, width, height);
    if (ret[3]) {
        throw takeFromExternrefTable0(ret[2]);
    }
    var v3 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
    return v3;
}
exports.wasmMaskMultiplyRgb = wasmMaskMultiplyRgb;

/**
 * `rgb`: row-major linear RGB in `[0,1]`, length `width * height * 3`.
 * `monocular_cue_blend`: forwarded to [`MeshDepthConfig::monocular_cue_blend`] (0 = off).
 * @param {Float32Array} rgb
 * @param {number} width
 * @param {number} height
 * @param {number} monocular_cue_blend
 * @returns {WasmMeshDepthStyleResult}
 */
function wasmMeshDepthConfidenceRgbF32(rgb, width, height, monocular_cue_blend) {
    const ptr0 = passArrayF32ToWasm0(rgb, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.wasmMeshDepthConfidenceRgbF32(ptr0, len0, width, height, monocular_cue_blend);
    if (ret[2]) {
        throw takeFromExternrefTable0(ret[1]);
    }
    return WasmMeshDepthStyleResult.__wrap(ret[0]);
}
exports.wasmMeshDepthConfidenceRgbF32 = wasmMeshDepthConfidenceRgbF32;

/**
 * Topology/depth-gated residual deblur over row-major linear RGB `[0,1]`.
 * @param {Float32Array} rgb
 * @param {number} width
 * @param {number} height
 * @param {number} overpotential
 * @param {number} admission_floor
 * @param {number} saturation_limit
 * @param {number} depth_confidence_blend
 * @returns {WasmMeshResidualDeblurResult}
 */
function wasmMeshResidualDeblurRgbF32(rgb, width, height, overpotential, admission_floor, saturation_limit, depth_confidence_blend) {
    const ptr0 = passArrayF32ToWasm0(rgb, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.wasmMeshResidualDeblurRgbF32(ptr0, len0, width, height, overpotential, admission_floor, saturation_limit, depth_confidence_blend);
    if (ret[2]) {
        throw takeFromExternrefTable0(ret[1]);
    }
    return WasmMeshResidualDeblurResult.__wrap(ret[0]);
}
exports.wasmMeshResidualDeblurRgbF32 = wasmMeshResidualDeblurRgbF32;

/**
 * **Monster LUT**: 66-Chord Topological Grading via WASM.
 * @param {Float32Array} rgb
 * @param {Float32Array} topology_alpha
 * @param {number} width
 * @param {number} height
 * @param {Float32Array} lut_66_weights
 * @returns {WasmTopologicalCalibrationResult}
 */
function wasmMonsterLutApply(rgb, topology_alpha, width, height, lut_66_weights) {
    const ptr0 = passArrayF32ToWasm0(rgb, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passArrayF32ToWasm0(topology_alpha, wasm.__wbindgen_malloc);
    const len1 = WASM_VECTOR_LEN;
    const ptr2 = passArrayF32ToWasm0(lut_66_weights, wasm.__wbindgen_malloc);
    const len2 = WASM_VECTOR_LEN;
    const ret = wasm.wasmMonsterLutApply(ptr0, len0, ptr1, len1, width, height, ptr2, len2);
    if (ret[2]) {
        throw takeFromExternrefTable0(ret[1]);
    }
    return WasmTopologicalCalibrationResult.__wrap(ret[0]);
}
exports.wasmMonsterLutApply = wasmMonsterLutApply;

/**
 * Richardson–Lucy line-PSF motion deblur over row-major linear RGB `[0,1]`.
 * @param {Float32Array} rgb
 * @param {number} width
 * @param {number} height
 * @param {number} angle_deg
 * @param {number} length_px
 * @param {number} iterations
 * @param {number} damping
 * @param {boolean} emit_layers
 * @returns {WasmMotionDeblurResult}
 */
function wasmMotionDeblurRgbF32(rgb, width, height, angle_deg, length_px, iterations, damping, emit_layers) {
    const ptr0 = passArrayF32ToWasm0(rgb, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.wasmMotionDeblurRgbF32(ptr0, len0, width, height, angle_deg, length_px, iterations, damping, emit_layers);
    if (ret[2]) {
        throw takeFromExternrefTable0(ret[1]);
    }
    return WasmMotionDeblurResult.__wrap(ret[0]);
}
exports.wasmMotionDeblurRgbF32 = wasmMotionDeblurRgbF32;

/**
 * @param {Float32Array} rgb
 * @param {Float32Array} plane
 * @param {Float32Array} topology_alpha
 * @param {number} width
 * @param {number} height
 * @param {number} light_x
 * @param {number} light_y
 * @param {number} light_z
 * @param {number} diffuse_strength
 * @param {number} ambient
 * @returns {Float32Array}
 */
function wasmRelightLambertMasked(rgb, plane, topology_alpha, width, height, light_x, light_y, light_z, diffuse_strength, ambient) {
    const ptr0 = passArrayF32ToWasm0(rgb, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passArrayF32ToWasm0(plane, wasm.__wbindgen_malloc);
    const len1 = WASM_VECTOR_LEN;
    const ptr2 = passArrayF32ToWasm0(topology_alpha, wasm.__wbindgen_malloc);
    const len2 = WASM_VECTOR_LEN;
    const ret = wasm.wasmRelightLambertMasked(ptr0, len0, ptr1, len1, ptr2, len2, width, height, light_x, light_y, light_z, diffuse_strength, ambient);
    if (ret[3]) {
        throw takeFromExternrefTable0(ret[2]);
    }
    var v4 = getArrayF32FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
    return v4;
}
exports.wasmRelightLambertMasked = wasmRelightLambertMasked;

/**
 * **Topological Recompositioning**: Relight using high-dimensional harmonic potential.
 * @param {Float32Array} rgb
 * @param {Float32Array} topology_alpha
 * @param {number} width
 * @param {number} height
 * @param {number} overpotential
 * @param {number} harmonic_threshold
 * @returns {WasmRelightMonsterMeshResult}
 */
function wasmRelightMonsterMeshMasked(rgb, topology_alpha, width, height, overpotential, harmonic_threshold) {
    const ptr0 = passArrayF32ToWasm0(rgb, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ptr1 = passArrayF32ToWasm0(topology_alpha, wasm.__wbindgen_malloc);
    const len1 = WASM_VECTOR_LEN;
    const ret = wasm.wasmRelightMonsterMeshMasked(ptr0, len0, ptr1, len1, width, height, overpotential, harmonic_threshold);
    if (ret[2]) {
        throw takeFromExternrefTable0(ret[1]);
    }
    return WasmRelightMonsterMeshResult.__wrap(ret[0]);
}
exports.wasmRelightMonsterMeshMasked = wasmRelightMonsterMeshMasked;

/**
 * `rgb`: row-major unit linear floats `width * height * 3`.
 * @param {Float32Array} rgb
 * @param {number} width
 * @param {number} height
 * @returns {WasmTopologyMaskResult}
 */
function wasmTopologyMaskBuildRgbF32(rgb, width, height) {
    const ptr0 = passArrayF32ToWasm0(rgb, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.wasmTopologyMaskBuildRgbF32(ptr0, len0, width, height);
    if (ret[2]) {
        throw takeFromExternrefTable0(ret[1]);
    }
    return WasmTopologyMaskResult.__wrap(ret[0]);
}
exports.wasmTopologyMaskBuildRgbF32 = wasmTopologyMaskBuildRgbF32;
function __wbg_get_imports() {
    const import0 = {
        __proto__: null,
        __wbg_Error_960c155d3d49e4c2: function(arg0, arg1) {
            const ret = Error(getStringFromWasm0(arg0, arg1));
            return ret;
        },
        __wbg_Number_32bf70a599af1d4b: function(arg0) {
            const ret = Number(arg0);
            return ret;
        },
        __wbg___wbindgen_boolean_get_6ea149f0a8dcc5ff: function(arg0) {
            const v = arg0;
            const ret = typeof(v) === 'boolean' ? v : undefined;
            return isLikeNone(ret) ? 0xFFFFFF : ret ? 1 : 0;
        },
        __wbg___wbindgen_debug_string_ab4b34d23d6778bd: function(arg0, arg1) {
            const ret = debugString(arg1);
            const ptr1 = passStringToWasm0(ret, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
            const len1 = WASM_VECTOR_LEN;
            getDataViewMemory0().setInt32(arg0 + 4 * 1, len1, true);
            getDataViewMemory0().setInt32(arg0 + 4 * 0, ptr1, true);
        },
        __wbg___wbindgen_in_a5d8b22e52b24dd1: function(arg0, arg1) {
            const ret = arg0 in arg1;
            return ret;
        },
        __wbg___wbindgen_is_function_3baa9db1a987f47d: function(arg0) {
            const ret = typeof(arg0) === 'function';
            return ret;
        },
        __wbg___wbindgen_is_object_63322ec0cd6ea4ef: function(arg0) {
            const val = arg0;
            const ret = typeof(val) === 'object' && val !== null;
            return ret;
        },
        __wbg___wbindgen_is_string_6df3bf7ef1164ed3: function(arg0) {
            const ret = typeof(arg0) === 'string';
            return ret;
        },
        __wbg___wbindgen_is_undefined_29a43b4d42920abd: function(arg0) {
            const ret = arg0 === undefined;
            return ret;
        },
        __wbg___wbindgen_jsval_loose_eq_cac3565e89b4134c: function(arg0, arg1) {
            const ret = arg0 == arg1;
            return ret;
        },
        __wbg___wbindgen_number_get_c7f42aed0525c451: function(arg0, arg1) {
            const obj = arg1;
            const ret = typeof(obj) === 'number' ? obj : undefined;
            getDataViewMemory0().setFloat64(arg0 + 8 * 1, isLikeNone(ret) ? 0 : ret, true);
            getDataViewMemory0().setInt32(arg0 + 4 * 0, !isLikeNone(ret), true);
        },
        __wbg___wbindgen_string_get_7ed5322991caaec5: function(arg0, arg1) {
            const obj = arg1;
            const ret = typeof(obj) === 'string' ? obj : undefined;
            var ptr1 = isLikeNone(ret) ? 0 : passStringToWasm0(ret, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
            var len1 = WASM_VECTOR_LEN;
            getDataViewMemory0().setInt32(arg0 + 4 * 1, len1, true);
            getDataViewMemory0().setInt32(arg0 + 4 * 0, ptr1, true);
        },
        __wbg___wbindgen_throw_6b64449b9b9ed33c: function(arg0, arg1) {
            throw new Error(getStringFromWasm0(arg0, arg1));
        },
        __wbg_call_14b169f759b26747: function() { return handleError(function (arg0, arg1) {
            const ret = arg0.call(arg1);
            return ret;
        }, arguments); },
        __wbg_call_a24592a6f349a97e: function() { return handleError(function (arg0, arg1, arg2) {
            const ret = arg0.call(arg1, arg2);
            return ret;
        }, arguments); },
        __wbg_crypto_38df2bab126b63dc: function(arg0) {
            const ret = arg0.crypto;
            return ret;
        },
        __wbg_done_9158f7cc8751ba32: function(arg0) {
            const ret = arg0.done;
            return ret;
        },
        __wbg_error_a6fa202b58aa1cd3: function(arg0, arg1) {
            let deferred0_0;
            let deferred0_1;
            try {
                deferred0_0 = arg0;
                deferred0_1 = arg1;
                console.error(getStringFromWasm0(arg0, arg1));
            } finally {
                wasm.__wbindgen_free(deferred0_0, deferred0_1, 1);
            }
        },
        __wbg_fill_58b126f99e178bb9: function(arg0, arg1, arg2, arg3) {
            const ret = arg0.fill(arg1, arg2 >>> 0, arg3 >>> 0);
            return ret;
        },
        __wbg_getRandomValues_c44a50d8cfdaebeb: function() { return handleError(function (arg0, arg1) {
            arg0.getRandomValues(arg1);
        }, arguments); },
        __wbg_get_1affdbdd5573b16a: function() { return handleError(function (arg0, arg1) {
            const ret = Reflect.get(arg0, arg1);
            return ret;
        }, arguments); },
        __wbg_get_unchecked_17f53dad852b9588: function(arg0, arg1) {
            const ret = arg0[arg1 >>> 0];
            return ret;
        },
        __wbg_get_with_ref_key_6412cf3094599694: function(arg0, arg1) {
            const ret = arg0[arg1];
            return ret;
        },
        __wbg_instanceof_ArrayBuffer_7c8433c6ed14ffe3: function(arg0) {
            let result;
            try {
                result = arg0 instanceof ArrayBuffer;
            } catch (_) {
                result = false;
            }
            const ret = result;
            return ret;
        },
        __wbg_instanceof_Uint8Array_152ba1f289edcf3f: function(arg0) {
            let result;
            try {
                result = arg0 instanceof Uint8Array;
            } catch (_) {
                result = false;
            }
            const ret = result;
            return ret;
        },
        __wbg_isArray_c3109d14ffc06469: function(arg0) {
            const ret = Array.isArray(arg0);
            return ret;
        },
        __wbg_isSafeInteger_4fc213d1989d6d2a: function(arg0) {
            const ret = Number.isSafeInteger(arg0);
            return ret;
        },
        __wbg_iterator_013bc09ec998c2a7: function() {
            const ret = Symbol.iterator;
            return ret;
        },
        __wbg_length_3d4ecd04bd8d22f1: function(arg0) {
            const ret = arg0.length;
            return ret;
        },
        __wbg_length_9f1775224cf1d815: function(arg0) {
            const ret = arg0.length;
            return ret;
        },
        __wbg_length_b5ce18c1fcafa728: function(arg0) {
            const ret = arg0.length;
            return ret;
        },
        __wbg_msCrypto_bd5a034af96bcba6: function(arg0) {
            const ret = arg0.msCrypto;
            return ret;
        },
        __wbg_new_0c7403db6e782f19: function(arg0) {
            const ret = new Uint8Array(arg0);
            return ret;
        },
        __wbg_new_227d7c05414eb861: function() {
            const ret = new Error();
            return ret;
        },
        __wbg_new_682678e2f47e32bc: function() {
            const ret = new Array();
            return ret;
        },
        __wbg_new_aa8d0fa9762c29bd: function() {
            const ret = new Object();
            return ret;
        },
        __wbg_new_with_length_8c854e41ea4dae9b: function(arg0) {
            const ret = new Uint8Array(arg0 >>> 0);
            return ret;
        },
        __wbg_next_0340c4ae324393c3: function() { return handleError(function (arg0) {
            const ret = arg0.next();
            return ret;
        }, arguments); },
        __wbg_next_7646edaa39458ef7: function(arg0) {
            const ret = arg0.next;
            return ret;
        },
        __wbg_node_84ea875411254db1: function(arg0) {
            const ret = arg0.node;
            return ret;
        },
        __wbg_process_44c7a14e11e9f69e: function(arg0) {
            const ret = arg0.process;
            return ret;
        },
        __wbg_prototypesetcall_a6b02eb00b0f4ce2: function(arg0, arg1, arg2) {
            Uint8Array.prototype.set.call(getArrayU8FromWasm0(arg0, arg1), arg2);
        },
        __wbg_push_471a5b068a5295f6: function(arg0, arg1) {
            const ret = arg0.push(arg1);
            return ret;
        },
        __wbg_randomFillSync_6c25eac9869eb53c: function() { return handleError(function (arg0, arg1) {
            arg0.randomFillSync(arg1);
        }, arguments); },
        __wbg_require_b4edbdcf3e2a1ef0: function() { return handleError(function () {
            const ret = module.require;
            return ret;
        }, arguments); },
        __wbg_set_3bf1de9fab0cd644: function(arg0, arg1, arg2) {
            arg0[arg1 >>> 0] = arg2;
        },
        __wbg_set_6be42768c690e380: function(arg0, arg1, arg2) {
            arg0[arg1] = arg2;
        },
        __wbg_set_c625d5a9e6b7554b: function(arg0, arg1, arg2) {
            arg0.set(getArrayF32FromWasm0(arg1, arg2));
        },
        __wbg_stack_3b0d974bbf31e44f: function(arg0, arg1) {
            const ret = arg1.stack;
            const ptr1 = passStringToWasm0(ret, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
            const len1 = WASM_VECTOR_LEN;
            getDataViewMemory0().setInt32(arg0 + 4 * 1, len1, true);
            getDataViewMemory0().setInt32(arg0 + 4 * 0, ptr1, true);
        },
        __wbg_static_accessor_GLOBAL_8cfadc87a297ca02: function() {
            const ret = typeof global === 'undefined' ? null : global;
            return isLikeNone(ret) ? 0 : addToExternrefTable0(ret);
        },
        __wbg_static_accessor_GLOBAL_THIS_602256ae5c8f42cf: function() {
            const ret = typeof globalThis === 'undefined' ? null : globalThis;
            return isLikeNone(ret) ? 0 : addToExternrefTable0(ret);
        },
        __wbg_static_accessor_SELF_e445c1c7484aecc3: function() {
            const ret = typeof self === 'undefined' ? null : self;
            return isLikeNone(ret) ? 0 : addToExternrefTable0(ret);
        },
        __wbg_static_accessor_WINDOW_f20e8576ef1e0f17: function() {
            const ret = typeof window === 'undefined' ? null : window;
            return isLikeNone(ret) ? 0 : addToExternrefTable0(ret);
        },
        __wbg_subarray_f8ca46a25b1f5e0d: function(arg0, arg1, arg2) {
            const ret = arg0.subarray(arg1 >>> 0, arg2 >>> 0);
            return ret;
        },
        __wbg_value_ee3a06f4579184fa: function(arg0) {
            const ret = arg0.value;
            return ret;
        },
        __wbg_versions_276b2795b1c6a219: function(arg0) {
            const ret = arg0.versions;
            return ret;
        },
        __wbindgen_cast_0000000000000001: function(arg0) {
            // Cast intrinsic for `F64 -> Externref`.
            const ret = arg0;
            return ret;
        },
        __wbindgen_cast_0000000000000002: function(arg0, arg1) {
            // Cast intrinsic for `Ref(Slice(U8)) -> NamedExternref("Uint8Array")`.
            const ret = getArrayU8FromWasm0(arg0, arg1);
            return ret;
        },
        __wbindgen_cast_0000000000000003: function(arg0, arg1) {
            // Cast intrinsic for `Ref(String) -> Externref`.
            const ret = getStringFromWasm0(arg0, arg1);
            return ret;
        },
        __wbindgen_init_externref_table: function() {
            const table = wasm.__wbindgen_externrefs;
            const offset = table.grow(4);
            table.set(0, undefined);
            table.set(offset + 0, undefined);
            table.set(offset + 1, null);
            table.set(offset + 2, true);
            table.set(offset + 3, false);
        },
    };
    return {
        __proto__: null,
        "./distributed_inference_bg.js": import0,
    };
}

const TensorBuilderFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_tensorbuilder_free(ptr >>> 0, 1));
const WasmGemma4PipelineFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_wasmgemma4pipeline_free(ptr >>> 0, 1));
const WasmHostBackendFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_wasmhostbackend_free(ptr >>> 0, 1));
const WasmLlamaPipelineFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_wasmllamapipeline_free(ptr >>> 0, 1));
const WasmMeshDepthStyleResultFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_wasmmeshdepthstyleresult_free(ptr >>> 0, 1));
const WasmMeshResidualDeblurResultFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_wasmmeshresidualdeblurresult_free(ptr >>> 0, 1));
const WasmMotionDeblurResultFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_wasmmotiondeblurresult_free(ptr >>> 0, 1));
const WasmPneumaTtsPipelineFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_wasmpneumattspipeline_free(ptr >>> 0, 1));
const WasmRelightMonsterMeshResultFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_wasmrelightmonstermeshresult_free(ptr >>> 0, 1));
const WasmSfmResultFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_wasmsfmresult_free(ptr >>> 0, 1));
const WasmTopologicalCalibrationResultFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_wasmtopologicalcalibrationresult_free(ptr >>> 0, 1));
const WasmTopologyMaskResultFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_wasmtopologymaskresult_free(ptr >>> 0, 1));
const WasmWhisperPipelineFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_wasmwhisperpipeline_free(ptr >>> 0, 1));

function addToExternrefTable0(obj) {
    const idx = wasm.__externref_table_alloc();
    wasm.__wbindgen_externrefs.set(idx, obj);
    return idx;
}

function _assertClass(instance, klass) {
    if (!(instance instanceof klass)) {
        throw new Error(`expected instance of ${klass.name}`);
    }
}

function debugString(val) {
    // primitive types
    const type = typeof val;
    if (type == 'number' || type == 'boolean' || val == null) {
        return  `${val}`;
    }
    if (type == 'string') {
        return `"${val}"`;
    }
    if (type == 'symbol') {
        const description = val.description;
        if (description == null) {
            return 'Symbol';
        } else {
            return `Symbol(${description})`;
        }
    }
    if (type == 'function') {
        const name = val.name;
        if (typeof name == 'string' && name.length > 0) {
            return `Function(${name})`;
        } else {
            return 'Function';
        }
    }
    // objects
    if (Array.isArray(val)) {
        const length = val.length;
        let debug = '[';
        if (length > 0) {
            debug += debugString(val[0]);
        }
        for(let i = 1; i < length; i++) {
            debug += ', ' + debugString(val[i]);
        }
        debug += ']';
        return debug;
    }
    // Test for built-in
    const builtInMatches = /\[object ([^\]]+)\]/.exec(toString.call(val));
    let className;
    if (builtInMatches && builtInMatches.length > 1) {
        className = builtInMatches[1];
    } else {
        // Failed to match the standard '[object ClassName]'
        return toString.call(val);
    }
    if (className == 'Object') {
        // we're a user defined class or Object
        // JSON.stringify avoids problems with cycles, and is generally much
        // easier than looping through ownProperties of `val`.
        try {
            return 'Object(' + JSON.stringify(val) + ')';
        } catch (_) {
            return 'Object';
        }
    }
    // errors
    if (val instanceof Error) {
        return `${val.name}: ${val.message}\n${val.stack}`;
    }
    // TODO we could test for more things here, like `Set`s and `Map`s.
    return className;
}

function getArrayF32FromWasm0(ptr, len) {
    ptr = ptr >>> 0;
    return getFloat32ArrayMemory0().subarray(ptr / 4, ptr / 4 + len);
}

function getArrayF64FromWasm0(ptr, len) {
    ptr = ptr >>> 0;
    return getFloat64ArrayMemory0().subarray(ptr / 8, ptr / 8 + len);
}

function getArrayI64FromWasm0(ptr, len) {
    ptr = ptr >>> 0;
    return getBigInt64ArrayMemory0().subarray(ptr / 8, ptr / 8 + len);
}

function getArrayU16FromWasm0(ptr, len) {
    ptr = ptr >>> 0;
    return getUint16ArrayMemory0().subarray(ptr / 2, ptr / 2 + len);
}

function getArrayU32FromWasm0(ptr, len) {
    ptr = ptr >>> 0;
    return getUint32ArrayMemory0().subarray(ptr / 4, ptr / 4 + len);
}

function getArrayU8FromWasm0(ptr, len) {
    ptr = ptr >>> 0;
    return getUint8ArrayMemory0().subarray(ptr / 1, ptr / 1 + len);
}

let cachedBigInt64ArrayMemory0 = null;
function getBigInt64ArrayMemory0() {
    if (cachedBigInt64ArrayMemory0 === null || cachedBigInt64ArrayMemory0.byteLength === 0) {
        cachedBigInt64ArrayMemory0 = new BigInt64Array(wasm.memory.buffer);
    }
    return cachedBigInt64ArrayMemory0;
}

let cachedBigUint64ArrayMemory0 = null;
function getBigUint64ArrayMemory0() {
    if (cachedBigUint64ArrayMemory0 === null || cachedBigUint64ArrayMemory0.byteLength === 0) {
        cachedBigUint64ArrayMemory0 = new BigUint64Array(wasm.memory.buffer);
    }
    return cachedBigUint64ArrayMemory0;
}

let cachedDataViewMemory0 = null;
function getDataViewMemory0() {
    if (cachedDataViewMemory0 === null || cachedDataViewMemory0.buffer.detached === true || (cachedDataViewMemory0.buffer.detached === undefined && cachedDataViewMemory0.buffer !== wasm.memory.buffer)) {
        cachedDataViewMemory0 = new DataView(wasm.memory.buffer);
    }
    return cachedDataViewMemory0;
}

let cachedFloat32ArrayMemory0 = null;
function getFloat32ArrayMemory0() {
    if (cachedFloat32ArrayMemory0 === null || cachedFloat32ArrayMemory0.byteLength === 0) {
        cachedFloat32ArrayMemory0 = new Float32Array(wasm.memory.buffer);
    }
    return cachedFloat32ArrayMemory0;
}

let cachedFloat64ArrayMemory0 = null;
function getFloat64ArrayMemory0() {
    if (cachedFloat64ArrayMemory0 === null || cachedFloat64ArrayMemory0.byteLength === 0) {
        cachedFloat64ArrayMemory0 = new Float64Array(wasm.memory.buffer);
    }
    return cachedFloat64ArrayMemory0;
}

function getStringFromWasm0(ptr, len) {
    ptr = ptr >>> 0;
    return decodeText(ptr, len);
}

let cachedUint16ArrayMemory0 = null;
function getUint16ArrayMemory0() {
    if (cachedUint16ArrayMemory0 === null || cachedUint16ArrayMemory0.byteLength === 0) {
        cachedUint16ArrayMemory0 = new Uint16Array(wasm.memory.buffer);
    }
    return cachedUint16ArrayMemory0;
}

let cachedUint32ArrayMemory0 = null;
function getUint32ArrayMemory0() {
    if (cachedUint32ArrayMemory0 === null || cachedUint32ArrayMemory0.byteLength === 0) {
        cachedUint32ArrayMemory0 = new Uint32Array(wasm.memory.buffer);
    }
    return cachedUint32ArrayMemory0;
}

let cachedUint8ArrayMemory0 = null;
function getUint8ArrayMemory0() {
    if (cachedUint8ArrayMemory0 === null || cachedUint8ArrayMemory0.byteLength === 0) {
        cachedUint8ArrayMemory0 = new Uint8Array(wasm.memory.buffer);
    }
    return cachedUint8ArrayMemory0;
}

function handleError(f, args) {
    try {
        return f.apply(this, args);
    } catch (e) {
        const idx = addToExternrefTable0(e);
        wasm.__wbindgen_exn_store(idx);
    }
}

function isLikeNone(x) {
    return x === undefined || x === null;
}

function passArray16ToWasm0(arg, malloc) {
    const ptr = malloc(arg.length * 2, 2) >>> 0;
    getUint16ArrayMemory0().set(arg, ptr / 2);
    WASM_VECTOR_LEN = arg.length;
    return ptr;
}

function passArray32ToWasm0(arg, malloc) {
    const ptr = malloc(arg.length * 4, 4) >>> 0;
    getUint32ArrayMemory0().set(arg, ptr / 4);
    WASM_VECTOR_LEN = arg.length;
    return ptr;
}

function passArray64ToWasm0(arg, malloc) {
    const ptr = malloc(arg.length * 8, 8) >>> 0;
    getBigUint64ArrayMemory0().set(arg, ptr / 8);
    WASM_VECTOR_LEN = arg.length;
    return ptr;
}

function passArray8ToWasm0(arg, malloc) {
    const ptr = malloc(arg.length * 1, 1) >>> 0;
    getUint8ArrayMemory0().set(arg, ptr / 1);
    WASM_VECTOR_LEN = arg.length;
    return ptr;
}

function passArrayF32ToWasm0(arg, malloc) {
    const ptr = malloc(arg.length * 4, 4) >>> 0;
    getFloat32ArrayMemory0().set(arg, ptr / 4);
    WASM_VECTOR_LEN = arg.length;
    return ptr;
}

function passArrayF64ToWasm0(arg, malloc) {
    const ptr = malloc(arg.length * 8, 8) >>> 0;
    getFloat64ArrayMemory0().set(arg, ptr / 8);
    WASM_VECTOR_LEN = arg.length;
    return ptr;
}

function passArrayJsValueToWasm0(array, malloc) {
    const ptr = malloc(array.length * 4, 4) >>> 0;
    for (let i = 0; i < array.length; i++) {
        const add = addToExternrefTable0(array[i]);
        getDataViewMemory0().setUint32(ptr + 4 * i, add, true);
    }
    WASM_VECTOR_LEN = array.length;
    return ptr;
}

function passStringToWasm0(arg, malloc, realloc) {
    if (realloc === undefined) {
        const buf = cachedTextEncoder.encode(arg);
        const ptr = malloc(buf.length, 1) >>> 0;
        getUint8ArrayMemory0().subarray(ptr, ptr + buf.length).set(buf);
        WASM_VECTOR_LEN = buf.length;
        return ptr;
    }

    let len = arg.length;
    let ptr = malloc(len, 1) >>> 0;

    const mem = getUint8ArrayMemory0();

    let offset = 0;

    for (; offset < len; offset++) {
        const code = arg.charCodeAt(offset);
        if (code > 0x7F) break;
        mem[ptr + offset] = code;
    }
    if (offset !== len) {
        if (offset !== 0) {
            arg = arg.slice(offset);
        }
        ptr = realloc(ptr, len, len = offset + arg.length * 3, 1) >>> 0;
        const view = getUint8ArrayMemory0().subarray(ptr + offset, ptr + len);
        const ret = cachedTextEncoder.encodeInto(arg, view);

        offset += ret.written;
        ptr = realloc(ptr, len, offset, 1) >>> 0;
    }

    WASM_VECTOR_LEN = offset;
    return ptr;
}

function takeFromExternrefTable0(idx) {
    const value = wasm.__wbindgen_externrefs.get(idx);
    wasm.__externref_table_dealloc(idx);
    return value;
}

let cachedTextDecoder = new TextDecoder('utf-8', { ignoreBOM: true, fatal: true });
cachedTextDecoder.decode();
function decodeText(ptr, len) {
    return cachedTextDecoder.decode(getUint8ArrayMemory0().subarray(ptr, ptr + len));
}

const cachedTextEncoder = new TextEncoder();

if (!('encodeInto' in cachedTextEncoder)) {
    cachedTextEncoder.encodeInto = function (arg, view) {
        const buf = cachedTextEncoder.encode(arg);
        view.set(buf);
        return {
            read: arg.length,
            written: buf.length
        };
    };
}

let WASM_VECTOR_LEN = 0;

const wasmPath = `${__dirname}/distributed_inference_bg.wasm`;
const wasmBytes = require('fs').readFileSync(wasmPath);
const wasmModule = new WebAssembly.Module(wasmBytes);
let wasm = new WebAssembly.Instance(wasmModule, __wbg_get_imports()).exports;
wasm.__wbindgen_start();
