// @a0n/gnosis/forkjoin-app/client/components/LiveWord/LiveWord.tsx
/**
 * LiveWord - Collaborative editable word in the hero section
 *
 * Features:
 * - Real-time collaborative editing
 * - AI-powered content moderation
 * - Resets to "Alive" every minute
 * - Shows countdown to next reset
 */

import { useState, useEffect, useRef, useCallback } from '@a0n/raect';
import { getLiveWordTraining } from './LiveWordTraining';

interface LiveWordState {
  word: string;
  lastReset: number;
  resetsIn: number;
  trainingExamples?: number;
}

const FOREGROUND_POLL_MS = 9000;
const BACKGROUND_POLL_MS = 30000;
const POLL_JITTER_MS = 2000;
const TRAINING_STATS_POLL_MS = 15000;

/**
 * Renders the Live Word view.
 */
export function LiveWord() {
  const [state, setState] = useState<LiveWordState>({
    word: 'Alive',
    lastReset: Date.now(),
    resetsIn: 60000,
  });
  const [isEditing, setIsEditing] = useState(false);
  const [inputValue, setInputValue] = useState('');
  const [error, setError] = useState<string | null>(null);
  const [isSubmitting, setIsSubmitting] = useState(false);
  const [errorExpanded, setErrorExpanded] = useState(false);
  const [trainingStats, setTrainingStats] = useState({
    examples: 0,
    ready: false,
  });
  const inputRef = useRef<HTMLInputElement>(null);
  const errorRef = useRef<HTMLSpanElement>(null);
  const pollTimeoutRef = useRef<ReturnType<typeof setTimeout> | null>(null);
  const trainingStatsIntervalRef = useRef<ReturnType<
    typeof setInterval
  > | null>(null);
  const fetchInFlightRef = useRef(false);

  // Focus error dialog when it appears (for keyboard accessibility)
  useEffect(() => {
    if (error && errorRef.current) {
      errorRef.current.focus();
    }
  }, [error]);

  // Initialize training service
  useEffect(() => {
    const training = getLiveWordTraining();
    let disposed = false;
    let initialized = false;
    let cancelDeferredInit: (() => void) | null = null;

    const updateTrainingStats = () => {
      const stats = training.getStats();
      setTrainingStats({
        examples: stats.examplesLoaded,
        ready: stats.modelReady,
      });
    };

    const initializeTraining = async () => {
      try {
        await training.initialize();
        if (disposed) return;
        initialized = true;
        updateTrainingStats();
      } catch {
        // Keep live word interactive even if training init fails.
      }
    };

    const browserWindow = window as Window & {
      requestIdleCallback?: (
        callback: (deadline: IdleDeadline) => void,
        options?: IdleRequestOptions
      ) => number;
      cancelIdleCallback?: (handle: number) => void;
    };

    if (typeof browserWindow.requestIdleCallback === 'function') {
      const idleHandle = browserWindow.requestIdleCallback(
        () => {
          void initializeTraining();
        },
        { timeout: 1500 }
      );
      cancelDeferredInit = () => {
        browserWindow.cancelIdleCallback?.(idleHandle);
      };
    } else {
      const timeoutHandle = setTimeout(() => {
        void initializeTraining();
      }, 750);
      cancelDeferredInit = () => clearTimeout(timeoutHandle);
    }

    // Keep this low frequency and avoid work in hidden tabs.
    trainingStatsIntervalRef.current = setInterval(() => {
      if (!initialized) return;
      if (document.visibilityState === 'hidden') return;
      updateTrainingStats();
    }, TRAINING_STATS_POLL_MS);

    return () => {
      disposed = true;
      cancelDeferredInit?.();
      if (trainingStatsIntervalRef.current) {
        clearInterval(trainingStatsIntervalRef.current);
      }
      training.destroy();
    };
  }, []);

  // Fetch current word
  const fetchWord = useCallback(async () => {
    if (fetchInFlightRef.current) return;
    fetchInFlightRef.current = true;

    try {
      const response = await fetch('/api/liveword', { cache: 'no-store' });
      if (response.ok) {
        const data = (await response.json()) as LiveWordState;
        setState((prev) => {
          if (
            prev.word === data.word &&
            prev.lastReset === data.lastReset &&
            prev.resetsIn === data.resetsIn &&
            prev.trainingExamples === data.trainingExamples
          ) {
            return prev;
          }
          return data;
        });
      }
    } catch {
      // Silently fail, keep current state
    } finally {
      fetchInFlightRef.current = false;
    }
  }, []);

  // Poll for updates with visibility-aware backoff and jitter.
  useEffect(() => {
    let active = true;

    const clearPollTimer = () => {
      if (pollTimeoutRef.current) {
        clearTimeout(pollTimeoutRef.current);
        pollTimeoutRef.current = null;
      }
    };

    const scheduleNextPoll = () => {
      if (!active) return;
      const baseInterval =
        document.visibilityState === 'visible'
          ? FOREGROUND_POLL_MS
          : BACKGROUND_POLL_MS;
      const jitterMs = Math.floor(Math.random() * POLL_JITTER_MS);
      pollTimeoutRef.current = setTimeout(async () => {
        await fetchWord();
        scheduleNextPoll();
      }, baseInterval + jitterMs);
    };

    void fetchWord().then(() => {
      scheduleNextPoll();
    });

    const handleVisibilityChange = () => {
      clearPollTimer();
      if (document.visibilityState === 'visible') {
        void fetchWord();
      }
      scheduleNextPoll();
    };

    document.addEventListener('visibilitychange', handleVisibilityChange);

    return () => {
      active = false;
      clearPollTimer();
      document.removeEventListener('visibilitychange', handleVisibilityChange);
    };
  }, [fetchWord]);

  // Update countdown timer
  useEffect(() => {
    const updateCountdown = () => {
      if (document.visibilityState === 'hidden') return;
      setState((prev) => {
        const nextResetsIn = Math.max(0, 60000 - (Date.now() - prev.lastReset));
        if (nextResetsIn === prev.resetsIn) {
          return prev;
        }
        return {
          ...prev,
          resetsIn: nextResetsIn,
        };
      });
    };

    const timer = setInterval(() => {
      updateCountdown();
    }, 1000);

    const handleVisibilityChange = () => {
      if (document.visibilityState === 'visible') {
        updateCountdown();
      }
    };

    document.addEventListener('visibilitychange', handleVisibilityChange);

    return () => {
      clearInterval(timer);
      document.removeEventListener('visibilitychange', handleVisibilityChange);
    };
  }, []);

  // Focus input when editing starts
  useEffect(() => {
    if (isEditing && inputRef.current) {
      inputRef.current.focus();
      inputRef.current.select();
    }
  }, [isEditing]);

  const handleClick = () => {
    if (!isEditing) {
      setInputValue(state.word);
      setIsEditing(true);
      setError(null);
      setErrorExpanded(false);
    }
  };

  const handleSubmit = async () => {
    if (isSubmitting) return;

    const trimmed = inputValue.trim();
    if (!trimmed) {
      setError('Cannot be empty');
      return;
    }
    if (trimmed.length > 50) {
      setError('Too long (max 50 chars)');
      return;
    }

    setIsSubmitting(true);
    setError(null);

    try {
      const response = await fetch('/api/liveword', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ word: trimmed }),
      });

      const data = (await response.json()) as LiveWordState & {
        error?: string;
        reason?: string;
        trainingExamples?: number;
      };

      if (!response.ok) {
        setError(data.reason || data.error || 'Failed to update');
        return;
      }

      // Add to training service
      const training = getLiveWordTraining();
      training.addExample(data.word);

      setState({
        word: data.word,
        lastReset: data.lastReset,
        resetsIn: data.resetsIn,
        trainingExamples: data.trainingExamples,
      });
      setIsEditing(false);
    } catch {
      setError('Network error');
    } finally {
      setIsSubmitting(false);
    }
  };

  const handleKeyDown = (e: React.KeyboardEvent) => {
    if (e.key === 'Enter') {
      e.preventDefault();
      handleSubmit();
    } else if (e.key === 'Escape') {
      setIsEditing(false);
      setError(null);
    }
  };

  const handleBlur = () => {
    // Small delay to allow click on submit button
    setTimeout(() => {
      if (!isSubmitting) {
        setIsEditing(false);
        setError(null);
      }
    }, 150);
  };

  const secondsUntilReset = Math.ceil(state.resetsIn / 1000);

  return (
    <span className="live-word-container">
      {/* Always render the display text to maintain layout */}
      <span
        className={`live-word-display ${isEditing ? 'editing' : ''}`}
        onClick={handleClick}
        role="button"
        tabIndex={isEditing ? -1 : 0}
        onKeyDown={(e) => e.key === 'Enter' && handleClick()}
        title={`Click to edit. Resets in ${secondsUntilReset}s`}
        aria-hidden={isEditing}
      >
        {isEditing ? inputValue || state.word : state.word}
        {!isEditing && (
          <span
            className="live-word-timer"
            aria-label={`Resets in ${secondsUntilReset} seconds`}
          >
            {secondsUntilReset}s
          </span>
        )}
        {!isEditing && trainingStats.ready && (
          <span
            className="live-word-neural"
            aria-label={`Neural model learning from ${trainingStats.examples} examples`}
          >
            neural
          </span>
        )}
      </span>

      {/* Overlay input when editing */}
      {isEditing && (
        <span className="live-word-editor">
          <input
            ref={inputRef}
            id="live-word-input"
            name="live-word"
            type="text"
            value={inputValue}
            onChange={(e) => setInputValue(e.target.value)}
            onKeyDown={handleKeyDown}
            onBlur={handleBlur}
            maxLength={50}
            className="live-word-input"
            disabled={isSubmitting}
            aria-label="Edit the live word"
          />
          {error && (
            <span
              ref={errorRef}
              className={`live-word-error ${errorExpanded ? 'expanded' : ''}`}
              onClick={(e) => {
                // Don't toggle if clicking the close button
                if ((e.target as HTMLElement).closest('.error-close')) return;
                setErrorExpanded(!errorExpanded);
              }}
              onKeyDown={(e) => {
                if (e.key === 'Escape') {
                  e.preventDefault();
                  setError(null);
                  setErrorExpanded(false);
                }
              }}
              role="dialog"
              aria-modal="true"
              aria-label="Content moderation message"
              tabIndex={0}
            >
              <span className="error-header">Nice try, but no.</span>
              <button
                className="error-close"
                onClick={() => {
                  setError(null);
                  setErrorExpanded(false);
                }}
                aria-label="Dismiss message"
                type="button"
              >
                ×
              </button>
              {errorExpanded && (
                <>
                  <span className="error-detail">{error}</span>
                  <span className="error-footer">
                    Thanks for testing our Edge Side Inference capabilities.
                    This is how AI-powered moderation will transform the web —
                    instant, on-device, and private.
                  </span>
                </>
              )}
              {!errorExpanded && (
                <span className="error-expand">(tap for details)</span>
              )}
            </span>
          )}
        </span>
      )}

      <style>{`
        .live-word-container {
          display: inline;
          position: relative;
        }

        .live-word-display {
          cursor: pointer;
          position: relative;
          transition: all 0.2s ease;
          border-bottom: 2px dashed rgba(16, 185, 129, 0.4);
          padding: 4px 12px 2px 12px;
          /* Override parent's transparent text fill to show the gradient */
          -webkit-text-fill-color: initial;
          background: linear-gradient(135deg, var(--color-accent, #10b981) 0%, #4488aa 100%);
          -webkit-background-clip: text;
          background-clip: text;
          color: transparent;
        }

        .live-word-display.editing {
          /* When editing, make the display invisible but keep layout */
          visibility: hidden;
        }

        .live-word-display:hover {
          filter: brightness(1.2);
          border-bottom-color: #10b981;
        }

        .live-word-display:hover .live-word-timer {
          opacity: 1;
        }

        .live-word-hint {
          color: inherit;
        }

        .live-word-timer {
          position: absolute;
          bottom: -1.8rem;
          right: -0.5rem;
          font-size: 0.7rem;
          font-weight: 500;
          color: #10b981;
          opacity: 0.7;
          transition: opacity 0.2s ease;
          font-family: var(--font-mono, monospace);
          -webkit-text-fill-color: #10b981;
          background: none;
          letter-spacing: 0.1em;
        }

        .live-word-neural {
          position: absolute;
          bottom: -1.8rem;
          left: -0.5rem;
          font-size: 0.6rem;
          font-weight: 500;
          color: #8b5cf6;
          opacity: 0.7;
          font-family: var(--font-mono, monospace);
          -webkit-text-fill-color: #8b5cf6;
          background: none;
          letter-spacing: 0.05em;
          animation: neural-pulse 2s ease-in-out infinite;
        }

        @keyframes neural-pulse {
          0%, 100% { opacity: 0.5; }
          50% { opacity: 0.9; }
        }

        .live-word-editor {
          position: absolute;
          left: -4px;
          top: -8px;
          right: -4px;
          bottom: -4px;
        }

        .live-word-input {
          position: absolute;
          left: 0;
          top: 0;
          width: 100%;
          height: 100%;
          font: inherit;
          font-size: inherit;
          font-weight: inherit;
          line-height: 1.2;
          color: #10b981;
          -webkit-text-fill-color: #10b981;
          background: rgba(16, 185, 129, 0.05);
          border: none;
          border-bottom: 2px solid #10b981;
          border-radius: 0;
          padding: 8px 16px 6px 16px;
          outline: none;
          text-align: center;
          box-sizing: border-box;
        }

        .live-word-input::selection {
          background: rgba(16, 185, 129, 0.3);
          color: #10b981;
          -webkit-text-fill-color: #10b981;
        }

        .live-word-input:disabled {
          opacity: 0.7;
        }

        .live-word-error {
          position: fixed;
          top: 50%;
          left: 50%;
          transform: translate(-50%, -50%);
          font-size: 1rem !important;
          line-height: 1.8 !important;
          letter-spacing: 0.05em !important;
          word-spacing: 0.1em !important;
          color: rgba(255, 255, 255, 0.9) !important;
          -webkit-text-fill-color: rgba(255, 255, 255, 0.9) !important;
          background: rgba(10, 10, 15, 0.98);
          padding: 28px 36px;
          border-radius: 12px;
          border: 1px solid rgba(16, 185, 129, 0.3);
          z-index: 10000;
          cursor: pointer;
          max-width: 480px;
          width: 90vw;
          box-shadow: 0 8px 32px rgba(0, 0, 0, 0.5), 0 0 60px rgba(16, 185, 129, 0.1);
          text-align: center;
          font-family: system-ui, -apple-system, sans-serif !important;
        }

        .live-word-error:hover {
          border-color: rgba(16, 185, 129, 0.5);
        }

        .live-word-error.expanded {
          text-align: left;
        }

        .error-header {
          display: block;
          font-weight: 600;
          font-size: 1.2rem !important;
          letter-spacing: 0.06em !important;
          color: #10b981 !important;
          -webkit-text-fill-color: #10b981 !important;
          background: none !important;
        }

        .error-expand {
          display: block;
          font-size: 0.85rem !important;
          letter-spacing: 0.05em !important;
          opacity: 0.6;
          margin-top: 10px;
          color: rgba(255, 255, 255, 0.7) !important;
          -webkit-text-fill-color: rgba(255, 255, 255, 0.7) !important;
          background: none !important;
        }

        .error-detail {
          display: block;
          margin-top: 18px;
          padding-top: 18px;
          border-top: 1px solid rgba(255, 255, 255, 0.1);
          font-size: 0.95rem !important;
          letter-spacing: 0.05em !important;
          color: rgba(255, 255, 255, 0.85) !important;
          -webkit-text-fill-color: rgba(255, 255, 255, 0.85) !important;
          background: none !important;
        }

        .error-footer {
          display: block;
          margin-top: 20px;
          padding-top: 18px;
          border-top: 1px solid rgba(16, 185, 129, 0.2);
          font-size: 0.85rem !important;
          font-style: italic;
          letter-spacing: 0.05em !important;
          color: rgba(16, 185, 129, 0.85) !important;
          -webkit-text-fill-color: rgba(16, 185, 129, 0.85) !important;
          background: none !important;
        }

        .error-close {
          position: absolute;
          top: 12px;
          right: 12px;
          width: 28px;
          height: 28px;
          display: flex;
          align-items: center;
          justify-content: center;
          font-size: 1.5rem;
          font-weight: 300;
          line-height: 1;
          color: rgba(255, 255, 255, 0.5);
          background: transparent;
          border: none;
          border-radius: 50%;
          cursor: pointer;
          transition: all 0.2s ease;
          padding: 0;
        }

        .error-close:hover {
          color: rgba(255, 255, 255, 0.9);
          background: rgba(255, 255, 255, 0.1);
        }

        .error-close:focus {
          outline: 2px solid rgba(16, 185, 129, 0.5);
          outline-offset: 2px;
        }
      `}</style>
    </span>
  );
}
