// @a0n/.ignored_gnosis/gnode/bridge-driver.test.ts
import { afterEach, beforeEach, describe, expect, it } from '@a0n/gnosis/test';
import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import {
  analyzeRuntimeModuleBraidFoldMetadata,
  runCli,
  selectRuntimeBraidFastPath,
} from './bridge-driver';

const RELAY_ENV_KEYS = [
  'GNODE_CACHE_AEON_RELAY_URL',
  'GNODE_CACHE_RELAY_URL',
  'GNODE_CACHE_AEON_RELAY_ROOM_PREFIX',
  'GNODE_CACHE_RELAY_ROOM_PREFIX',
  'GNODE_CACHE_AEON_RELAY_API_KEY',
  'GNODE_CACHE_RELAY_API_KEY',
  'GNODE_CACHE_AEON_RELAY_CLIENT_ID',
  'GNODE_CACHE_RELAY_CLIENT_ID',
  'GNODE_CACHE_AEON_RELAY_PROTOCOL',
  'GNODE_CACHE_RELAY_PROTOCOL',
  'GNODE_CACHE_AEON_RELAY_MODE',
  'GNODE_CACHE_RELAY_MODE',
  'GNODE_CACHE_AEON_RELAY_PRODUCT',
  'GNODE_CACHE_RELAY_PRODUCT',
  'GNODE_CACHE_AEON_RELAY_TIMEOUT_MS',
  'GNODE_CACHE_RELAY_TIMEOUT_MS',
] as const;

function createTempBridgeWorkspace(sourceText?: string): {
  readonly root: string;
  readonly appPath: string;
  readonly cacheDir: string;
} {
  const root = fs.mkdtempSync(path.join(os.tmpdir(), 'gnode-bridge-driver-'));
  const appPath = path.join(root, 'app.ts');
  const cacheDir = path.join(root, 'cache');

  fs.writeFileSync(
    appPath,
    sourceText ??
      [
        'function internalStep(input: number) {',
        '  return input + 2;',
        '}',
        '',
        'export function app(input: number) {',
        '  return internalStep(input);',
        '}',
        '',
      ].join('\n'),
    'utf8'
  );

  return { root, appPath, cacheDir };
}

function collectRelativeFiles(root: string): string[] {
  if (!fs.existsSync(root)) {
    return [];
  }

  const files: string[] = [];
  const queue = [root];

  while (queue.length > 0) {
    const current = queue.pop();
    if (!current) {
      continue;
    }

    for (const entry of fs.readdirSync(current, { withFileTypes: true })) {
      const entryPath = path.join(current, entry.name);
      if (entry.isDirectory()) {
        queue.push(entryPath);
        continue;
      }
      files.push(path.relative(root, entryPath));
    }
  }

  return files.sort();
}

function readCachedArtifactRecord(cacheDir: string): {
  readonly envelope: { readonly envelopeId: string; readonly digest: string };
} | null {
  const artifactFile = collectRelativeFiles(cacheDir).find((file) =>
    file.endsWith('artifact.qdoc')
  );
  if (!artifactFile) {
    return null;
  }

  const bytes = fs.readFileSync(path.join(cacheDir, artifactFile));
  const delta = JSON.parse(bytes.toString('utf8')) as {
    readonly edges?: ReadonlyArray<{
      readonly type?: string;
      readonly properties?: Readonly<Record<string, string>>;
    }>;
  };
  const observeEdge = [...(delta.edges ?? [])]
    .reverse()
    .find((edge) => edge.type === 'OBSERVE');
  const serializedValue = observeEdge?.properties?.value;
  if (!serializedValue) {
    return null;
  }

  const wrapped = JSON.parse(serializedValue) as {
    readonly value?: unknown;
  };
  return (
    (wrapped.value as {
      readonly envelope: {
        readonly envelopeId: string;
        readonly digest: string;
      };
    }) ?? null
  );
}

function captureStream<T>(
  stream: NodeJS.WriteStream,
  work: () => Promise<T>
): Promise<{ readonly result: T; readonly output: string }> {
  const originalWrite = stream.write.bind(stream);
  let output = '';

  stream.write = ((chunk: string | Uint8Array) => {
    output +=
      typeof chunk === 'string' ? chunk : Buffer.from(chunk).toString('utf8');
    return true;
  }) as typeof stream.write;

  return work().then(
    (result) => {
      stream.write = originalWrite as typeof stream.write;
      return { result, output };
    },
    (error) => {
      stream.write = originalWrite as typeof stream.write;
      throw error;
    }
  );
}

class FakeRelayWebSocket {
  static readonly OPEN = 1;
  static readonly CLOSED = 3;
  static instances: FakeRelayWebSocket[] = [];

  readonly url: string;
  readyState = 0;
  readonly sent: unknown[] = [];
  onopen: (() => void) | null = null;
  onmessage: ((event: { readonly data: string | Uint8Array }) => void) | null =
    null;
  onerror: (() => void) | null = null;
  onclose: (() => void) | null = null;

  constructor(url: string) {
    this.url = url;
    FakeRelayWebSocket.instances.push(this);
    queueMicrotask(() => {
      this.readyState = FakeRelayWebSocket.OPEN;
      this.onopen?.();
    });
  }

  send(payload: string | Uint8Array): void {
    this.sent.push(payload);

    if (typeof payload !== 'string') {
      return;
    }

    let envelope: { readonly type?: string } | null = null;
    try {
      envelope = JSON.parse(payload) as { readonly type?: string };
    } catch {
      envelope = null;
    }

    if (envelope?.type === 'join') {
      queueMicrotask(() => {
        this.onmessage?.({
          data: JSON.stringify({
            type: 'snapshot',
            serverSeq: 1,
          }),
        });
      });
    }
  }

  close(): void {
    this.readyState = FakeRelayWebSocket.CLOSED;
    this.onclose?.();
  }
}

describe('gnode bridge-driver cache persistence', () => {
  let previousEnv: Map<string, string | undefined>;
  let previousWebSocket: typeof globalThis.WebSocket | undefined;

  beforeEach(() => {
    previousEnv = new Map(
      ['GNODE_CACHE_DIR', ...RELAY_ENV_KEYS].map((key) => [
        key,
        process.env[key],
      ])
    );
    previousWebSocket = globalThis.WebSocket;
    FakeRelayWebSocket.instances = [];
  });

  afterEach(() => {
    for (const [key, value] of previousEnv) {
      if (value === undefined) {
        delete process.env[key];
        continue;
      }
      process.env[key] = value;
    }

    globalThis.WebSocket = previousWebSocket;
  });

  it('stores artifact and head records as qdoc updates instead of json files', async () => {
    const workspace = createTempBridgeWorkspace();
    process.env.GNODE_CACHE_DIR = workspace.cacheDir;

    try {
      const firstRun = await captureStream(process.stderr, async () =>
        captureStream(process.stdout, async () =>
          runCli(['compile', workspace.appPath, '--export', 'app'])
        )
      );
      expect(firstRun.result.result).toBe(0);

      const files = collectRelativeFiles(workspace.cacheDir);
      expect(files.some((file) => file.endsWith('artifact.qdoc'))).toBe(true);
      expect(files.some((file) => file.endsWith('.json'))).toBe(false);
      const firstEnvelope = readCachedArtifactRecord(
        workspace.cacheDir
      )?.envelope;
      expect(firstEnvelope?.envelopeId).toHaveLength(64);

      const secondRun = await captureStream(process.stderr, async () =>
        captureStream(process.stdout, async () =>
          runCli([
            'compile',
            workspace.appPath,
            '--export',
            'app',
            '--trace-timings',
          ])
        )
      );
      expect(secondRun.result.result).toBe(0);
      expect(secondRun.output).toContain('"cacheStatus": "hit"');
      const secondEnvelope = readCachedArtifactRecord(
        workspace.cacheDir
      )?.envelope;
      expect(secondEnvelope?.envelopeId).toBe(firstEnvelope?.envelopeId);
      expect(secondEnvelope?.digest).toBe(firstEnvelope?.digest);
    } finally {
      fs.rmSync(workspace.root, { recursive: true, force: true });
    }
  }, 15000);

  it('connects cache records to DashRelay-compatible sync only when relay env is present', async () => {
    const workspace = createTempBridgeWorkspace();
    process.env.GNODE_CACHE_DIR = workspace.cacheDir;
    process.env.GNODE_CACHE_RELAY_URL = 'wss://relay.example.invalid/cache';
    process.env.GNODE_CACHE_RELAY_ROOM_PREFIX = 'gnode-test-cache';
    globalThis.WebSocket =
      FakeRelayWebSocket as unknown as typeof globalThis.WebSocket;

    try {
      const result = await captureStream(process.stdout, async () =>
        runCli(['compile', workspace.appPath, '--export', 'app'])
      );
      expect(result.result).toBe(0);
      expect(FakeRelayWebSocket.instances.length).toBeGreaterThan(0);

      const sentFrames = FakeRelayWebSocket.instances.flatMap(
        (instance) => instance.sent
      );
      const joinFrames = sentFrames.filter(
        (payload): payload is string => typeof payload === 'string'
      );

      expect(
        joinFrames.some((payload) =>
          payload.includes('"room":"gnode-test-cache/artifact/')
        )
      ).toBe(true);
      expect(sentFrames.some((payload) => payload instanceof Uint8Array)).toBe(
        true
      );
    } finally {
      fs.rmSync(workspace.root, { recursive: true, force: true });
    }
  });

  it('forwards remaining CLI args as bridge input when --input-json is omitted', async () => {
    const workspace = createTempBridgeWorkspace(
      [
        'export function main(argv: string[]) {',
        '  return argv[0];',
        '}',
        '',
      ].join('\n')
    );

    try {
      const captured = await captureStream(process.stdout, async () =>
        runCli(['run', workspace.appPath, '--export', 'main', 'hello-cli'])
      );
      expect(captured.result).toBe(0);
      expect(captured.output).toContain('hello-cli');
    } finally {
      fs.rmSync(workspace.root, { recursive: true, force: true });
    }
  });

  it('runs bridge entrypoints with member call targets through the CLI path', async () => {
    const workspace = createTempBridgeWorkspace(
      [
        'export function app(input: string) {',
        '  console.log(`log:${input}`);',
        '  return { upper: input.toUpperCase() };',
        '}',
        '',
      ].join('\n')
    );

    try {
      const result = await captureStream(process.stderr, async () =>
        captureStream(process.stdout, async () =>
          runCli([
            'run',
            workspace.appPath,
            '--export',
            'app',
            '--input-json',
            JSON.stringify('ada'),
          ])
        )
      );

      expect(result.result.result).toBe(0);
      expect(result.output).toContain('[gnode run fallback]');
    } finally {
      fs.rmSync(workspace.root, { recursive: true, force: true });
    }
  });

  it('audits braid fold metadata without rewriting runtime module source', () => {
    const source = [
      '// @gnosis-braid-fold {"cycle":2,"gateCoordinates":[10,12,17,22],"saturationTrace":[10,30,90],"closedArchitecture":true,"endpointsLinked":true,"baseReady":true}',
      '// @gnosis-decomposition-triple {"cycle":2,"coordinates":[7,17,22]}',
      '// @gnosis-braid-packet {"cycle":2,"actionTrace":["fork","race","fold","vent"],"portCoordinates":[10,12,17,22],"saturationTrace":[10,30,90],"informationMassTrace":[120,360,1080],"leakageTrace":[0,18,78],"totalInformationMass":1560,"totalLeakage":96}',
      '// @gnosis-extended-aeon-transition {"ranges":[[1,12],[10,22],[20,32],[30,43],[50,55]],"keystoneTrace":[[7,11],[17,22],[27,32],[37,43],[55,55]],"primeTransition":{"aeon":"fourth","linearTerminal":42,"clinamen":1,"terminal":43,"foldKeystone":37,"ventKeystone":43},"closure":{"coordinate":55,"carrier":11,"cycles":5}}',
      '// @gnosis-interfere-port {"host":"gnosis-uring","portAction":"interfere","portCoordinate":43,"closureCoordinate":55,"terminal":true}',
      '// @gnosis-pressure-weighted-interference {"portAction":"interfere","anchorCoordinate":43,"closureCoordinate":55,"strideTrace":[33,31,26,21],"leakageTrace":[0,18,78],"weightedPressureTrace":[0,558,2028],"weightedPressureTotal":2586,"pressureUnit":"SPa","weighting":"stride-leakage-zip"}',
      '// @gnosis-closed-loop-runtime-metadata {"cycle":2,"closedTrace":[10,12,17,22,43,55,10],"centerlineTrace":[10,12,17,22,43],"representativeTrace":[180,180],"reconstructedTrace":[360,1080],"axisCoordinate":43,"closureCoordinate":55,"totalMass":1560,"totalLeakage":96,"pressureWeightedTotal":2586,"bucketRatio":130,"checksum":6439}',
      'export const untouched = 122;',
      '',
    ].join('\n');

    const trace = analyzeRuntimeModuleBraidFoldMetadata(source);

    expect(trace.status).toBe('hit');
    expect(trace.source).toBe('runtime-module');
    expect(trace.anchorCoordinates).toEqual([10, 17, 22]);
    expect(trace.certificate?.primeDoubleCrossingCount).toBe(5);
    expect(trace.decomposition.status).toBe('hit');
    expect(trace.decomposition.coordinates).toEqual([7, 17, 22]);
    expect(trace.decomposition.certificate?.primeDoubleCrossingCount).toBe(5);
    expect(trace.braidPacket.status).toBe('hit');
    expect(trace.braidPacket.cycle).toBe(2);
    expect(trace.braidPacket.certificate?.totalInformationMass).toBe(1560);
    expect(trace.extendedAeon.status).toBe('hit');
    expect(trace.extendedAeon.certificate?.primeTransition.terminal).toBe(43);
    expect(trace.extendedAeon.interfere.status).toBe('hit');
    expect(trace.extendedAeon.interfere.certificate?.portAction).toBe(
      'interfere'
    );
    expect(trace.extendedAeon.interfere.certificate?.portCoordinate).toBe(43);
    expect(trace.pressureWeightedInterference.status).toBe('hit');
    expect(trace.pressureWeightedInterference.weightedPressureTrace).toEqual([
      0, 558, 2028,
    ]);
    expect(trace.pressureWeightedInterference.weightedPressureTotal).toBe(2586);
    expect(
      trace.pressureWeightedInterference.certificate?.theoremIds
    ).toContain(
      'GamutBraid.exact_pressure_weighted_interference_anchor_packet_certified'
    );
    expect(trace.closedLoopRuntimeMetadata.status).toBe('hit');
    expect(trace.closedLoopRuntimeMetadata.closedTrace).toEqual([
      10, 12, 17, 22, 43, 55, 10,
    ]);
    expect(trace.closedLoopRuntimeMetadata.representativeTrace).toEqual([
      180, 180,
    ]);
    expect(trace.closedLoopRuntimeMetadata.reconstructedTrace).toEqual([
      360, 1080,
    ]);
    expect(trace.closedLoopRuntimeMetadata.checksum).toBe(6439);
    expect(trace.closedLoopRuntimeMetadata.certificate?.checksum).toBe(6439);
    expect(trace.closedLoopRuntimeMetadata.certificate?.theoremIds).toContain(
      'GamutBraid.exact_closed_loop_runtime_metadata_certified'
    );
    const fastPath = selectRuntimeBraidFastPath(trace);
    expect(fastPath.selected).toBe(true);
    expect(fastPath.path?.collapsedBridge).toEqual({
      from: 17,
      to: 22,
      stride: 5,
    });
    expect(source).toContain('export const untouched = 122;');
  });

  it('fails closed when decomposition metadata lacks braid closure', () => {
    const source = [
      '// @gnosis-decomposition-triple {"cycle":2,"coordinates":[7,17,22]}',
      'export const untouched = 122;',
      '',
    ].join('\n');

    const trace = analyzeRuntimeModuleBraidFoldMetadata(source);

    expect(trace.status).toBe('fallback');
    expect(trace.decomposition.status).toBe('fallback');
    expect(trace.decomposition.reason).toBe('braid fold certificate is absent');
    expect(trace.decomposition.certificate).toBe(null);
    expect(trace.braidPacket.status).toBe('fallback');
    expect(trace.extendedAeon.status).toBe('fallback');
    expect(trace.extendedAeon.interfere.status).toBe('fallback');
    expect(trace.pressureWeightedInterference.status).toBe('fallback');
    expect(trace.closedLoopRuntimeMetadata.status).toBe('fallback');
    expect(selectRuntimeBraidFastPath(trace).selected).toBe(false);
  });

  it('fails closed when braid fold metadata is missing an anchor', () => {
    const source = [
      '// @gnosis-braid-fold {"cycle":2,"gateCoordinates":[10,12,17],"saturationTrace":[10,30,90],"closedArchitecture":true,"endpointsLinked":true,"baseReady":true}',
      'export const untouched = 122;',
      '',
    ].join('\n');

    const trace = analyzeRuntimeModuleBraidFoldMetadata(source);

    expect(trace.status).toBe('fallback');
    expect(trace.anchorCoordinates).toEqual([10, 17]);
    expect(trace.certificate).toBe(null);
    expect(trace.braidPacket.status).toBe('fallback');
    expect(trace.extendedAeon.status).toBe('fallback');
    expect(trace.pressureWeightedInterference.status).toBe('fallback');
    expect(trace.closedLoopRuntimeMetadata.status).toBe('fallback');
    expect(selectRuntimeBraidFastPath(trace).selected).toBe(false);
  });
});
