// @a0n/gnosis/forkjoin-app/client/pages/AegisPage.tsx
/**
 * AegisPage - Dedicated page for Aegis identity/auth layer
 *
 * Decentralized authentication with UCAN and zero-knowledge proofs.
 */

import {
  ArrowLeft,
  Shield,
  Lock,
  Eye,
  Check,
  Ticket,
  Key,
  Fingerprint,
  Github,
  BookOpen,
  Code,
} from '@a0n/aeon-icons';
import { getFoundationDocsTrackUrl } from '../docsLinks';

/**
 * Renders the Aegis Page view.
 */
export function AegisPage() {
  return (
    <div className="product-page aegis-page">
      <nav className="product-nav">
        <a href="/" className="back-link">
          <ArrowLeft size={20} />
          <span>Back to Aeon Foundation</span>
        </a>
      </nav>

      <main className="product-content">
        <header className="product-header">
          <Shield size={64} className="product-icon" />
          <span className="product-badge">Identity Layer</span>
          <h1>Aegis</h1>
          <p className="product-tagline">
            Decentralized authentication that respects privacy. UCAN
            capabilities. Zero-knowledge proofs. No passwords.
          </p>
          <div className="product-actions">
            <a
              href="https://github.com/forkjoin-ai/aegis"
              className="button button-primary"
              target="_blank"
              rel="noopener noreferrer"
            >
              <Github size={18} />
              View on GitHub
            </a>
            <a
              href={getFoundationDocsTrackUrl('aegis')}
              className="button button-secondary"
              target="_blank"
              rel="noopener noreferrer"
            >
              <BookOpen size={18} />
              Documentation
            </a>
          </div>
        </header>

        <section className="product-section">
          <h2>Installation</h2>
          <pre className="install-code">npm install @affectively/aegis</pre>
        </section>

        <section className="product-section">
          <h2>Core Concepts</h2>
          <div className="concepts-grid">
            <div className="concept-card ucan-card">
              <Lock size={48} className="concept-icon" />
              <h3>UCAN</h3>
              <p className="concept-tagline">
                User Controlled Authorization Networks
              </p>
              <ul className="concept-features">
                <li>
                  <Check size={16} /> No central authority
                </li>
                <li>
                  <Check size={16} /> Offline-first verification
                </li>
                <li>
                  <Check size={16} /> Delegatable capabilities
                </li>
                <li>
                  <Check size={16} /> Anonymous by default
                </li>
                <li>
                  <Check size={16} /> Revocable grants
                </li>
              </ul>
            </div>

            <div className="concept-card zk-card">
              <Eye size={48} className="concept-icon" />
              <h3>Zero-Knowledge</h3>
              <p className="concept-tagline">Prove without revealing</p>
              <ul className="concept-features">
                <li>
                  <Check size={16} /> Prove age without birthday
                </li>
                <li>
                  <Check size={16} /> Prove membership without ID
                </li>
                <li>
                  <Check size={16} /> Prove ownership without keys
                </li>
                <li>
                  <Check size={16} /> Prove location without GPS
                </li>
                <li>
                  <Check size={16} /> Prove humanity without face
                </li>
              </ul>
            </div>
          </div>
        </section>

        <section className="product-section">
          <h2>How It Works</h2>
          <div className="zk-flow">
            <div className="zk-step">
              <Key size={32} className="step-icon" />
              <span className="step-label">Prover</span>
              <div className="step-value">knows secret</div>
            </div>
            <div className="zk-arrow">→</div>
            <div className="zk-step">
              <Fingerprint size={32} className="step-icon" />
              <span className="step-label">Proof</span>
              <div className="step-value">π = f(secret)</div>
            </div>
            <div className="zk-arrow">→</div>
            <div className="zk-step">
              <Check size={32} className="step-icon" />
              <span className="step-label">Verifier</span>
              <div className="step-value">learns nothing</div>
            </div>
          </div>
        </section>

        <section className="product-section">
          <h2>Live Demo</h2>
          <div className="demo-block">
            <h4>You're Already Using It</h4>
            <p>
              Right now, you have an anonymous UCAN grant stored in your
              browser. It lets you broadcast your cursor to everyone on this
              page — without signing up, without a password, without revealing
              who you are.
            </p>
            <div className="grant-badge">
              <Ticket size={20} className="grant-icon" />
              <span className="grant-text">Anonymous UCAN Active</span>
              <Check size={16} className="grant-check" />
            </div>
          </div>
        </section>

        <section className="product-section">
          <h2>Quick Start</h2>
          <div className="code-example">
            <div className="code-header">
              <Code size={18} />
              <span>Anonymous Identity</span>
            </div>
            <pre className="code-block">{`import { Aegis } from '@affectively/aegis';

// Generate anonymous identity
const keypair = await crypto.subtle.generateKey(
  { name: 'Ed25519' },
  true,
  ['sign', 'verify']
);

// Create UCAN with capabilities
const aegis = new Aegis();
const ucan = await aegis.createUCAN({
  issuer: keypair,
  audience: 'did:web:aeonflux.dev',
  capabilities: [
    { can: 'presence/broadcast' },
    { can: 'state/read' },
    { can: 'state/write' },
  ],
  expiration: inOneHour(),
});

// Store for later use
localStorage.setItem('ucan', ucan.encoded);`}</pre>
          </div>
        </section>

        <section className="product-section">
          <h2>Advanced Usage</h2>
          <div className="code-example">
            <div className="code-header">
              <Code size={18} />
              <span>Capability Delegation</span>
            </div>
            <pre className="code-block">{`import { Aegis, Capability } from '@affectively/aegis';

const aegis = new Aegis();

// Root UCAN with full access
const rootUcan = await aegis.createUCAN({
  issuer: adminKeypair,
  audience: 'did:key:user123',
  capabilities: [
    { can: 'document/*', with: 'doc:abc' },
  ],
});

// Delegate subset of capabilities
const delegatedUcan = await aegis.delegate({
  parent: rootUcan,
  issuer: userKeypair,
  audience: 'did:key:collaborator456',
  capabilities: [
    { can: 'document/read', with: 'doc:abc' },
    { can: 'document/comment', with: 'doc:abc' },
    // Cannot delegate 'document/delete' - not in parent
  ],
  expiration: inOneDay(),
});

// Verify capability chain
const valid = await aegis.verify(delegatedUcan);
console.log(valid.capabilities); // ['document/read', 'document/comment']`}</pre>
          </div>

          <div className="code-example">
            <div className="code-header">
              <Code size={18} />
              <span>Zero-Knowledge Proofs</span>
            </div>
            <pre className="code-block">{`import { Aegis, ZKProof } from '@affectively/aegis';

const aegis = new Aegis();

// Prove age without revealing birthday
const ageProof = await aegis.zk.prove({
  claim: 'age >= 18',
  secret: { birthday: '1990-05-15' },
  circuit: 'age-verification',
});

// Verify proof (learns nothing about actual birthday)
const valid = await aegis.zk.verify(ageProof);
console.log(valid); // true

// Prove membership without revealing identity
const membershipProof = await aegis.zk.prove({
  claim: 'isMember(group:premium)',
  secret: { memberId: 'user123' },
  circuit: 'set-membership',
});

// Prove ownership without revealing private key
const ownershipProof = await aegis.zk.prove({
  claim: 'owns(nft:abc)',
  secret: { privateKey: userPrivateKey },
  circuit: 'ownership',
});`}</pre>
          </div>

          <div className="code-example">
            <div className="code-header">
              <Code size={18} />
              <span>Revocation</span>
            </div>
            <pre className="code-block">{`import { Aegis } from '@affectively/aegis';

const aegis = new Aegis({
  revocationStore: env.KV, // Cloudflare KV for revocation list
});

// Revoke a UCAN
await aegis.revoke(ucan.cid, {
  reason: 'User requested revocation',
  revokedAt: Date.now(),
});

// Check if revoked before accepting
const status = await aegis.checkRevocation(ucan.cid);
if (status.revoked) {
  throw new Error(\`Token revoked: \${status.reason}\`);
}

// Revocation propagates to all delegated tokens
// Any token derived from this one is now invalid`}</pre>
          </div>
        </section>

        <section className="product-section">
          <h2>Resources</h2>
          <div className="resources-grid">
            <a
              href="https://github.com/forkjoin-ai/aegis"
              className="resource-card"
              target="_blank"
              rel="noopener noreferrer"
            >
              <Github size={24} />
              <h4>GitHub</h4>
              <p>Source code, issues, and discussions</p>
            </a>
            <a
              href={getFoundationDocsTrackUrl('aegis')}
              className="resource-card"
              target="_blank"
              rel="noopener noreferrer"
            >
              <BookOpen size={24} />
              <h4>Documentation</h4>
              <p>Full API reference and guides</p>
            </a>
            <a
              href="https://ucan.xyz"
              className="resource-card"
              target="_blank"
              rel="noopener noreferrer"
            >
              <Lock size={24} />
              <h4>UCAN Spec</h4>
              <p>Learn more about UCAN standard</p>
            </a>
          </div>
        </section>
      </main>

      <style>{`
        .aegis-page {
          --product-color: #ec4899;
          --product-color-secondary: #336699;
          min-height: 100vh;
          background: var(--color-bg);
        }

        .product-nav {
          padding: var(--space-lg) var(--space-xl);
          border-bottom: 1px solid var(--color-bg-tertiary);
        }

        .back-link {
          display: inline-flex;
          align-items: center;
          gap: var(--space-sm);
          color: var(--color-text-secondary);
          text-decoration: none;
          transition: color 0.2s ease;
        }

        .back-link:hover {
          color: var(--color-text);
        }

        .product-content {
          max-width: 900px;
          margin: 0 auto;
          padding: var(--space-3xl) var(--space-xl);
        }

        .product-header {
          text-align: center;
          margin-bottom: var(--space-3xl);
        }

        .product-icon {
          display: block;
          margin: 0 auto var(--space-lg);
          color: var(--product-color);
        }

        .product-badge {
          display: inline-block;
          padding: 0.25rem 0.75rem;
          font-size: 0.75rem;
          font-weight: 500;
          text-transform: uppercase;
          letter-spacing: 0.1em;
          color: var(--product-color);
          border: 1px solid var(--product-color);
          border-radius: var(--radius-full);
          margin-bottom: var(--space-md);
        }

        .product-header h1 {
          font-size: 3rem;
          margin-bottom: var(--space-md);
        }

        .product-tagline {
          font-size: 1.25rem;
          color: var(--color-text-secondary);
          max-width: 600px;
          margin: 0 auto var(--space-xl);
        }

        .product-actions {
          display: flex;
          gap: var(--space-md);
          justify-content: center;
          flex-wrap: wrap;
        }

        .product-actions .button {
          display: inline-flex;
          align-items: center;
          gap: var(--space-sm);
        }

        .product-section {
          margin-bottom: var(--space-3xl);
        }

        .product-section h2 {
          font-size: 1.5rem;
          margin-bottom: var(--space-lg);
          padding-bottom: var(--space-sm);
          border-bottom: 1px solid var(--color-bg-tertiary);
        }

        .install-code {
          font-family: var(--font-mono);
          font-size: 1.1rem;
          padding: var(--space-lg) var(--space-xl);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
          color: var(--product-color);
        }

        .concepts-grid {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(350px, 1fr));
          gap: var(--space-xl);
        }

        .concept-card {
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 16px;
          padding: var(--space-xl);
        }

        .ucan-card {
          border-color: rgba(236, 72, 153, 0.3);
        }

        .zk-card {
          border-color: rgba(51, 102, 153, 0.3);
        }

        .concept-icon {
          margin-bottom: var(--space-md);
        }

        .ucan-card .concept-icon {
          color: var(--product-color);
        }

        .zk-card .concept-icon {
          color: var(--product-color-secondary);
        }

        .concept-card h3 {
          margin-bottom: var(--space-xs);
        }

        .ucan-card h3 {
          color: var(--product-color);
        }

        .zk-card h3 {
          color: var(--product-color-secondary);
        }

        .concept-tagline {
          font-size: 0.9rem;
          color: var(--color-text-muted);
          margin-bottom: var(--space-lg);
        }

        .concept-features {
          list-style: none;
          padding: 0;
          margin: 0;
        }

        .concept-features li {
          display: flex;
          align-items: center;
          gap: var(--space-sm);
          padding: var(--space-xs) 0;
          color: var(--color-text-secondary);
        }

        .concept-features li svg {
          color: #10b981;
        }

        .zk-flow {
          display: flex;
          align-items: center;
          justify-content: center;
          gap: var(--space-xl);
          padding: var(--space-xl);
          background: var(--color-bg-secondary);
          border-radius: 16px;
          flex-wrap: wrap;
        }

        .zk-step {
          text-align: center;
        }

        .step-icon {
          color: var(--product-color-secondary);
          margin-bottom: var(--space-sm);
        }

        .step-label {
          display: block;
          font-size: 0.7rem;
          text-transform: uppercase;
          letter-spacing: 0.1em;
          color: var(--color-text-muted);
          margin-bottom: var(--space-xs);
        }

        .step-value {
          font-family: var(--font-mono);
          font-size: 0.9rem;
          color: var(--product-color-secondary);
        }

        .zk-arrow {
          font-size: 1.5rem;
          color: var(--color-text-muted);
        }

        .demo-block {
          text-align: center;
          padding: var(--space-xl);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 16px;
        }

        .demo-block h4 {
          color: var(--color-accent);
          margin-bottom: var(--space-md);
        }

        .demo-block p {
          color: var(--color-text-secondary);
          margin-bottom: var(--space-lg);
          max-width: 500px;
          margin-left: auto;
          margin-right: auto;
        }

        .grant-badge {
          display: inline-flex;
          align-items: center;
          gap: var(--space-sm);
          padding: var(--space-sm) var(--space-lg);
          background: rgba(16, 185, 129, 0.1);
          border: 1px solid rgba(16, 185, 129, 0.3);
          border-radius: 999px;
        }

        .grant-icon {
          color: #10b981;
        }

        .grant-text {
          font-size: 0.9rem;
        }

        .grant-check {
          color: #10b981;
        }

        .code-example {
          margin-bottom: var(--space-xl);
        }

        .code-header {
          display: flex;
          align-items: center;
          gap: var(--space-sm);
          margin-bottom: var(--space-sm);
          color: var(--color-text-secondary);
          font-size: 0.9rem;
        }

        .code-header svg {
          color: var(--product-color);
        }

        .code-block {
          font-family: var(--font-mono);
          font-size: 0.85rem;
          line-height: 1.6;
          padding: var(--space-lg);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
          overflow-x: auto;
          margin: 0;
        }

        .resources-grid {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(250px, 1fr));
          gap: var(--space-lg);
        }

        .resource-card {
          padding: var(--space-xl);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
          text-decoration: none;
          color: inherit;
          transition: all 0.2s ease;
        }

        .resource-card:hover {
          border-color: var(--product-color);
          transform: translateY(-2px);
        }

        .resource-card svg {
          color: var(--product-color);
          margin-bottom: var(--space-md);
        }

        .resource-card h4 {
          margin-bottom: var(--space-xs);
        }

        .resource-card p {
          color: var(--color-text-secondary);
          font-size: 0.85rem;
          margin: 0;
        }

        @media (max-width: 768px) {
          .concepts-grid {
            grid-template-columns: 1fr;
          }

          .zk-flow {
            flex-direction: column;
          }

          .zk-arrow {
            transform: rotate(90deg);
          }
        }
      `}</style>
    </div>
  );
}
