// @a0n/gnosis/forkjoin-app/client/pages/AeonPage.tsx
/**
 * AeonPage - Dedicated page for Aeon sync library
 *
 * Full documentation, code examples, and API reference.
 */

import {
  ArrowLeft,
  RefreshCw,
  GitCommit,
  Swords,
  Users,
  WifiOff,
  PackageOpen,
  Github,
  BookOpen,
  Code,
  Lock,
  Coins,
  Globe,
  Zap,
  Shield,
} from '@a0n/aeon-icons';
import { getFoundationDocsTrackUrl } from '../docsLinks';

/**
 * Renders the Aeon Page view.
 */
export function AeonPage() {
  return (
    <div className="product-page aeon-page">
      <nav className="product-nav">
        <a href="/" className="back-link">
          <ArrowLeft size={20} />
          <span>Back to Aeon Foundation</span>
        </a>
      </nav>

      <main className="product-content">
        <header className="product-header">
          <RefreshCw size={64} className="product-icon" />
          <span className="product-badge">The Sync Standard</span>
          <h1>Aeon</h1>
          <p className="product-tagline">
            Distributed synchronization, schema versioning, and conflict
            resolution. The collaborative primitives the web deserves.
          </p>
          <div className="product-actions">
            <a
              href="https://github.com/forkjoin-ai/aeon"
              className="button button-primary"
              target="_blank"
              rel="noopener noreferrer"
            >
              <Github size={18} />
              View on GitHub
            </a>
            <a
              href={getFoundationDocsTrackUrl('aeon')}
              className="button button-secondary"
              target="_blank"
              rel="noopener noreferrer"
            >
              <BookOpen size={18} />
              Documentation
            </a>
          </div>
        </header>

        <section className="product-section">
          <h2>Installation</h2>
          <pre className="install-code">npm install @a0n/aeon</pre>
        </section>

        <section className="product-section">
          <h2>Philosophy</h2>
          <blockquote className="philosophy-quote">
            "In Gnosticism, Aeons are divine bridges between realms. In your
            app, they bridge users to shared state."
          </blockquote>
        </section>

        <section className="product-section">
          <h2>Core Features</h2>
          <div className="features-grid">
            <FeatureCard
              icon={<RefreshCw size={32} />}
              title="Distributed Sync"
              description="Coordinate sync sessions across nodes. No central server required. State flows everywhere."
            />
            <FeatureCard
              icon={<GitCommit size={32} />}
              title="Schema Versioning"
              description="Evolve your data with confidence. Migrations, rollbacks, and audit trails built in."
            />
            <FeatureCard
              icon={<Swords size={32} />}
              title="Conflict Resolution"
              description="Last-write-wins, vector clocks, or majority vote. Pick your strategy per field."
            />
            <FeatureCard
              icon={<Users size={32} />}
              title="Real-time Presence"
              description="Who's here? What are they doing? Track cursors, selections, and activity across the network."
            />
            <FeatureCard
              icon={<WifiOff size={32} />}
              title="Offline-First"
              description="Queue operations when offline. Sync when back. AES-256-GCM encryption at rest."
            />
            <FeatureCard
              icon={<PackageOpen size={32} />}
              title="Delta Compression"
              description="70-90% payload reduction. Send only what changed. Adaptive compression based on network."
            />
          </div>
        </section>

        <section className="product-section">
          <h2>Quick Start</h2>
          <div className="code-example">
            <div className="code-header">
              <Code size={18} />
              <span>Basic Setup</span>
            </div>
            <pre className="code-block">{`import { SyncCoordinator, StateReconciler } from '@a0n/aeon';

// Create a sync coordinator
const coordinator = new SyncCoordinator();

// Register nodes in your cluster
coordinator.registerNode({ id: 'node-1', status: 'online' });
coordinator.registerNode({ id: 'node-2', status: 'online' });

// Create a sync session
const session = coordinator.createSyncSession('node-1', ['node-2']);

// Handle conflicts with your preferred strategy
const reconciler = new StateReconciler();
if (reconciler.detectConflicts('user:123')) {
  const result = reconciler.reconcileLastWriteWins(versions);
  console.log('Resolved:', result.mergedState);
}`}</pre>
          </div>
        </section>

        <section className="product-section">
          <h2>Advanced Usage</h2>
          <div className="code-example">
            <div className="code-header">
              <Code size={18} />
              <span>Schema Versioning</span>
            </div>
            <pre className="code-block">{`import { SchemaRegistry, MigrationRunner } from '@a0n/aeon';

// Define your schema with version
const registry = new SchemaRegistry();
registry.register({
  name: 'user',
  version: 2,
  fields: {
    id: { type: 'string', required: true },
    name: { type: 'string', required: true },
    email: { type: 'string', required: true },
    preferences: { type: 'object', default: {} },
  },
});

// Run migrations automatically
const runner = new MigrationRunner(registry);
await runner.migrate('user', 1, 2, (doc) => ({
  ...doc,
  preferences: doc.settings ?? {}, // Rename field
}));`}</pre>
          </div>

          <div className="code-example">
            <div className="code-header">
              <Code size={18} />
              <span>Real-time Presence</span>
            </div>
            <pre className="code-block">{`import { PresenceManager } from '@a0n/aeon';

const presence = new PresenceManager({
  roomId: 'document:123',
  throttleMs: 50,
});

// Announce your presence
presence.join({
  userId: 'user-abc',
  cursor: { x: 0, y: 0 },
  selection: null,
});

// Update cursor position
document.onmousemove = (e) => {
  presence.update({
    cursor: { x: e.clientX, y: e.clientY },
  });
};

// Listen for other users
presence.on('change', (users) => {
  renderCursors(users);
});`}</pre>
          </div>

          <div className="code-example">
            <div className="code-header">
              <Code size={18} />
              <span>Conflict Resolution Strategies</span>
            </div>
            <pre className="code-block">{`import { StateReconciler, ConflictStrategy } from '@a0n/aeon';

const reconciler = new StateReconciler();

// Last-write-wins (default)
reconciler.setStrategy('title', ConflictStrategy.LAST_WRITE_WINS);

// Vector clock ordering
reconciler.setStrategy('content', ConflictStrategy.VECTOR_CLOCK);

// Majority vote (requires 3+ nodes)
reconciler.setStrategy('status', ConflictStrategy.MAJORITY_VOTE);

// Custom merge function
reconciler.setStrategy('tags', ConflictStrategy.CUSTOM, {
  merge: (local, remote) => [...new Set([...local, ...remote])],
});`}</pre>
          </div>
        </section>

        <section className="product-section">
          <h2>Comparison</h2>
          <div className="comparison-table">
            <table>
              <thead>
                <tr>
                  <th>Feature</th>
                  <th>Aeon</th>
                  <th>Yjs</th>
                  <th>Automerge</th>
                </tr>
              </thead>
              <tbody>
                <tr>
                  <td>Schema Versioning</td>
                  <td className="yes">Yes</td>
                  <td className="no">No</td>
                  <td className="no">No</td>
                </tr>
                <tr>
                  <td>Migrations & Rollback</td>
                  <td className="yes">Yes</td>
                  <td className="no">No</td>
                  <td className="no">No</td>
                </tr>
                <tr>
                  <td>Replication Policies</td>
                  <td className="yes">Yes</td>
                  <td className="no">No</td>
                  <td className="no">No</td>
                </tr>
                <tr>
                  <td>Multiple Merge Strategies</td>
                  <td className="yes">Yes</td>
                  <td className="partial">Partial</td>
                  <td className="partial">Partial</td>
                </tr>
                <tr>
                  <td>TypeScript-first</td>
                  <td className="yes">Yes</td>
                  <td className="partial">Partial</td>
                  <td className="partial">Partial</td>
                </tr>
                <tr>
                  <td>Zero Dependencies*</td>
                  <td className="yes">Yes</td>
                  <td className="no">No</td>
                  <td className="no">No</td>
                </tr>
              </tbody>
            </table>
            <p className="table-note">*Only eventemitter3 for events</p>
          </div>
        </section>

        <section className="product-section compute-pool-section">
          <h2>Contribute to the Network</h2>
          <p className="section-intro">
            Join thousands of browsers powering the world's most private AI
            network.
          </p>

          <div className="contribution-features">
            <div className="contribution-card">
              <Lock size={24} />
              <h4>100% Private</h4>
              <p>
                Inference runs in your browser using WebGPU or WASM. Your
                prompts never touch our servers.
              </p>
            </div>
            <div className="contribution-card">
              <Coins size={24} />
              <h4>Earn Karma</h4>
              <p>
                Contribute idle compute, earn karma. Unlock premium tools and
                DAO voting rights.
              </p>
            </div>
            <div className="contribution-card">
              <Globe size={24} />
              <h4>House Fleet</h4>
              <p>
                Join 10,000+ browsers in the distributed network. Faster than
                cloud APIs, cheaper than GPUs.
              </p>
            </div>
          </div>

          <div className="technical-details">
            <div className="detail-card">
              <div className="detail-header">
                <Zap size={20} />
                <h4>GPU Acceleration</h4>
              </div>
              <ul className="detail-list">
                <li>
                  <span className="dot dot-green" /> WebNN: Native hardware
                  acceleration (NPU/GPU)
                </li>
                <li>
                  <span className="dot dot-green" /> WebGPU: Modern GPU compute
                  shaders
                </li>
                <li>
                  <span className="dot dot-blue" /> CPU: WASM fallback for all
                  devices
                </li>
              </ul>
            </div>
            <div className="detail-card">
              <div className="detail-header">
                <Shield size={20} />
                <h4>Privacy Guarantees</h4>
              </div>
              <ul className="detail-list">
                <li>
                  <span className="dot dot-blue" /> All inference runs locally
                  in your browser
                </li>
                <li>
                  <span className="dot dot-blue" /> No data sent to external
                  servers
                </li>
                <li>
                  <span className="dot dot-blue" /> Contribution pauses when tab
                  is hidden
                </li>
              </ul>
            </div>
          </div>

          <div className="contribution-cta">
            <h4>Start Contributing</h4>
            <p>Enable the toggle to join the network</p>
            <div className="placeholder-toggle">
              [Compute Pool Toggle Component]
            </div>
          </div>
        </section>

        <section className="product-section">
          <h2>Resources</h2>
          <div className="resources-grid">
            <a
              href="https://github.com/forkjoin-ai/aeon"
              className="resource-card"
              target="_blank"
              rel="noopener noreferrer"
            >
              <Github size={24} />
              <h4>GitHub</h4>
              <p>Source code, issues, and discussions</p>
            </a>
            <a
              href={getFoundationDocsTrackUrl('aeon')}
              className="resource-card"
              target="_blank"
              rel="noopener noreferrer"
            >
              <BookOpen size={24} />
              <h4>Documentation</h4>
              <p>Full API reference and guides</p>
            </a>
            <a
              href="https://github.com/forkjoin-ai/aeon/tree/main/examples"
              className="resource-card"
              target="_blank"
              rel="noopener noreferrer"
            >
              <Code size={24} />
              <h4>Examples</h4>
              <p>Sample projects and use cases</p>
            </a>
          </div>
        </section>
      </main>

      <style>{`
        .aeon-page {
          --product-color: #336699;
          min-height: 100vh;
          background: var(--color-bg);
        }

        .product-nav {
          padding: var(--space-lg) var(--space-xl);
          border-bottom: 1px solid var(--color-bg-tertiary);
        }

        .back-link {
          display: inline-flex;
          align-items: center;
          gap: var(--space-sm);
          color: var(--color-text-secondary);
          text-decoration: none;
          transition: color 0.2s ease;
        }

        .back-link:hover {
          color: var(--color-text);
        }

        .product-content {
          max-width: 900px;
          margin: 0 auto;
          padding: var(--space-3xl) var(--space-xl);
        }

        .product-header {
          text-align: center;
          margin-bottom: var(--space-3xl);
        }

        .product-icon {
          display: block;
          margin: 0 auto var(--space-lg);
          color: var(--product-color);
        }

        .product-badge {
          display: inline-block;
          padding: 0.25rem 0.75rem;
          font-size: 0.75rem;
          font-weight: 500;
          text-transform: uppercase;
          letter-spacing: 0.1em;
          color: var(--product-color);
          border: 1px solid var(--product-color);
          border-radius: var(--radius-full);
          margin-bottom: var(--space-md);
        }

        .product-header h1 {
          font-size: 3rem;
          margin-bottom: var(--space-md);
        }

        .product-tagline {
          font-size: 1.25rem;
          color: var(--color-text-secondary);
          max-width: 600px;
          margin: 0 auto var(--space-xl);
        }

        .product-actions {
          display: flex;
          gap: var(--space-md);
          justify-content: center;
          flex-wrap: wrap;
        }

        .product-actions .button {
          display: inline-flex;
          align-items: center;
          gap: var(--space-sm);
        }

        .product-section {
          margin-bottom: var(--space-3xl);
        }

        .product-section h2 {
          font-size: 1.5rem;
          margin-bottom: var(--space-lg);
          padding-bottom: var(--space-sm);
          border-bottom: 1px solid var(--color-bg-tertiary);
        }

        .install-code {
          font-family: var(--font-mono);
          font-size: 1.1rem;
          padding: var(--space-lg) var(--space-xl);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
          color: var(--product-color);
        }

        .philosophy-quote {
          font-size: 1.25rem;
          font-style: italic;
          color: var(--color-text-secondary);
          padding: var(--space-xl);
          border-left: 4px solid var(--product-color);
          background: var(--color-bg-secondary);
          border-radius: 0 12px 12px 0;
          margin: 0;
        }

        .features-grid {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(280px, 1fr));
          gap: var(--space-lg);
        }

        .feature-card {
          padding: var(--space-xl);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
        }

        .feature-card svg {
          color: var(--product-color);
          margin-bottom: var(--space-md);
        }

        .feature-card h4 {
          margin-bottom: var(--space-sm);
        }

        .feature-card p {
          color: var(--color-text-secondary);
          font-size: 0.9rem;
          margin: 0;
        }

        .code-example {
          margin-bottom: var(--space-xl);
        }

        .code-header {
          display: flex;
          align-items: center;
          gap: var(--space-sm);
          margin-bottom: var(--space-sm);
          color: var(--color-text-secondary);
          font-size: 0.9rem;
        }

        .code-header svg {
          color: var(--product-color);
        }

        .code-block {
          font-family: var(--font-mono);
          font-size: 0.85rem;
          line-height: 1.6;
          padding: var(--space-lg);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
          overflow-x: auto;
          margin: 0;
        }

        .comparison-table table {
          width: 100%;
          border-collapse: collapse;
          background: var(--color-bg-secondary);
          border-radius: 12px;
          overflow: hidden;
        }

        .comparison-table th,
        .comparison-table td {
          padding: var(--space-md);
          text-align: left;
          border-bottom: 1px solid var(--color-bg-tertiary);
        }

        .comparison-table th {
          background: var(--color-bg-tertiary);
          font-weight: 600;
        }

        .comparison-table .yes {
          color: var(--color-success);
        }

        .comparison-table .no {
          color: var(--color-text-muted);
        }

        .comparison-table .partial {
          color: var(--color-warning);
        }

        .table-note {
          font-size: 0.75rem;
          color: var(--color-text-muted);
          margin-top: var(--space-sm);
        }

        .resources-grid {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(250px, 1fr));
          gap: var(--space-lg);
        }

        .resource-card {
          padding: var(--space-xl);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
          text-decoration: none;
          color: inherit;
          transition: all 0.2s ease;
        }

        .resource-card:hover {
          border-color: var(--product-color);
          transform: translateY(-2px);
        }

        .resource-card svg {
          color: var(--product-color);
          margin-bottom: var(--space-md);
        }

        .resource-card h4 {
          margin-bottom: var(--space-xs);
        }

        .resource-card p {
          color: var(--color-text-secondary);
          font-size: 0.85rem;
          margin: 0;
        }

        .compute-pool-section .section-intro {
          color: var(--color-text-secondary);
          font-size: 1.1rem;
          margin-bottom: var(--space-xl);
        }

        .contribution-features {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(240px, 1fr));
          gap: var(--space-lg);
          margin-bottom: var(--space-xl);
        }

        .contribution-card {
          padding: var(--space-lg);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
        }

        .contribution-card svg {
          color: var(--product-color);
          margin-bottom: var(--space-sm);
        }

        .contribution-card h4 {
          margin-bottom: var(--space-xs);
          font-size: 1rem;
        }

        .contribution-card p {
          color: var(--color-text-secondary);
          font-size: 0.85rem;
          margin: 0;
          line-height: 1.5;
        }

        .technical-details {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(300px, 1fr));
          gap: var(--space-lg);
          margin-bottom: var(--space-xl);
        }

        .detail-card {
          padding: var(--space-lg);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
        }

        .detail-header {
          display: flex;
          align-items: center;
          gap: var(--space-sm);
          margin-bottom: var(--space-md);
        }

        .detail-header svg {
          color: var(--product-color);
        }

        .detail-header h4 {
          margin: 0;
          font-size: 1rem;
        }

        .detail-list {
          list-style: none;
          padding: 0;
          margin: 0;
        }

        .detail-list li {
          display: flex;
          align-items: center;
          gap: var(--space-sm);
          color: var(--color-text-secondary);
          font-size: 0.85rem;
          margin-bottom: var(--space-xs);
        }

        .dot {
          width: 6px;
          height: 6px;
          border-radius: 50%;
          flex-shrink: 0;
        }

        .dot-green {
          background: #10b981;
        }

        .dot-blue {
          background: #3b82f6;
        }

        .contribution-cta {
          text-align: center;
          padding: var(--space-xl);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
        }

        .contribution-cta h4 {
          margin-bottom: var(--space-xs);
          font-size: 1.25rem;
        }

        .contribution-cta p {
          color: var(--color-text-secondary);
          margin-bottom: var(--space-lg);
        }
      `}</style>
    </div>
  );
}

interface FeatureCardProps {
  icon: React.ReactNode;
  title: string;
  description: string;
}

function FeatureCard({ icon, title, description }: FeatureCardProps) {
  return (
    <div className="feature-card">
      {icon}
      <h4>{title}</h4>
      <p>{description}</p>
    </div>
  );
}
