// @a0n/gnosis/forkjoin-app/client/pages/Cookies.tsx
/**
 * Cookie Policy Page
 */

import { ArrowLeft, Cookie, X, Check } from '@a0n/aeon-icons';

/**
 * Renders the Cookies view.
 */
export function Cookies() {
  return (
    <div className="legal-page">
      <nav className="legal-nav">
        <a href="/" className="back-link">
          <ArrowLeft size={20} />
          <span>Back to Aeon Foundation</span>
        </a>
      </nav>

      <main className="legal-content">
        <div className="legal-header">
          <Cookie size={48} className="legal-icon" />
          <h1>Cookie Policy</h1>
          <p className="legal-date">Last updated: February 2026</p>
        </div>

        <section className="legal-section highlight-section">
          <h2>TL;DR: We Don't Use Tracking Cookies</h2>
          <p>
            We don't track you. We don't use analytics cookies. We don't use
            third-party cookies. We don't fingerprint your browser. We believe
            the web should respect privacy by default.
          </p>
        </section>

        <section className="legal-section">
          <h2>What We Don't Use</h2>
          <div className="cookie-list">
            <div className="cookie-item no">
              <X size={20} />
              <div>
                <strong>Analytics Cookies</strong>
                <span>No Google Analytics, Mixpanel, or similar tracking</span>
              </div>
            </div>
            <div className="cookie-item no">
              <X size={20} />
              <div>
                <strong>Advertising Cookies</strong>
                <span>We don't show ads or track you for advertising</span>
              </div>
            </div>
            <div className="cookie-item no">
              <X size={20} />
              <div>
                <strong>Third-Party Cookies</strong>
                <span>
                  No Facebook Pixel, Twitter, or other external trackers
                </span>
              </div>
            </div>
            <div className="cookie-item no">
              <X size={20} />
              <div>
                <strong>Cross-Site Tracking</strong>
                <span>We don't follow you across the internet</span>
              </div>
            </div>
          </div>
        </section>

        <section className="legal-section">
          <h2>What We Do Use</h2>
          <div className="cookie-list">
            <div className="cookie-item yes">
              <Check size={20} />
              <div>
                <strong>Session Storage</strong>
                <span>
                  Anonymous session ID for real-time presence (deleted on close)
                </span>
              </div>
            </div>
            <div className="cookie-item yes">
              <Check size={20} />
              <div>
                <strong>Local Storage</strong>
                <span>Your anonymous UCAN grant (never sent to servers)</span>
              </div>
            </div>
            <div className="cookie-item yes">
              <Check size={20} />
              <div>
                <strong>Theme Preference</strong>
                <span>Light/dark mode setting (stored locally)</span>
              </div>
            </div>
          </div>
        </section>

        <section className="legal-section">
          <h2>Technical Details</h2>
          <table className="cookie-table">
            <thead>
              <tr>
                <th>Name</th>
                <th>Purpose</th>
                <th>Duration</th>
                <th>Type</th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  <code>aeon_session</code>
                </td>
                <td>Anonymous presence session</td>
                <td>Session</td>
                <td>sessionStorage</td>
              </tr>
              <tr>
                <td>
                  <code>aeon_ucan</code>
                </td>
                <td>Anonymous capability grant</td>
                <td>1 hour</td>
                <td>localStorage</td>
              </tr>
              <tr>
                <td>
                  <code>aeon_theme</code>
                </td>
                <td>Color scheme preference</td>
                <td>1 year</td>
                <td>localStorage</td>
              </tr>
            </tbody>
          </table>
        </section>

        <section className="legal-section">
          <h2>Your Rights</h2>
          <p>
            Since we use localStorage and sessionStorage (not cookies), no
            consent banner is legally required. However, you can always:
          </p>
          <ul>
            <li>Clear your browser's local storage to reset all data</li>
            <li>Use private/incognito mode for a fresh session</li>
            <li>
              Block JavaScript entirely (site still works, just without
              real-time features)
            </li>
          </ul>
        </section>

        <section className="legal-section">
          <h2>Contact</h2>
          <p>
            Questions? Email us at{' '}
            <a href="mailto:privacy@aeonflux.dev">privacy@aeonflux.dev</a>
          </p>
        </section>
      </main>

      <style>{`
        .legal-page {
          min-height: 100vh;
          background: var(--color-bg);
          color: var(--color-text);
        }

        .legal-nav {
          padding: var(--space-lg) var(--space-xl);
          border-bottom: 1px solid var(--color-bg-tertiary);
        }

        .back-link {
          display: inline-flex;
          align-items: center;
          gap: var(--space-sm);
          color: var(--color-text-secondary);
          text-decoration: none;
          font-size: 0.9rem;
          transition: color 0.2s ease;
        }

        .back-link:hover {
          color: var(--color-text);
        }

        .legal-content {
          max-width: 800px;
          margin: 0 auto;
          padding: var(--space-3xl) var(--space-xl);
        }

        .legal-header {
          text-align: center;
          margin-bottom: var(--space-3xl);
        }

        .legal-icon {
          color: var(--color-accent);
          margin-bottom: var(--space-md);
        }

        .legal-header h1 {
          margin-bottom: var(--space-sm);
        }

        .legal-date {
          font-size: 0.875rem;
          color: var(--color-text-muted);
        }

        .legal-section {
          margin-bottom: var(--space-3xl);
        }

        .highlight-section {
          padding: var(--space-xl);
          background: rgba(16, 185, 129, 0.1);
          border: 1px solid rgba(16, 185, 129, 0.3);
          border-radius: 16px;
        }

        .highlight-section h2 {
          color: #10b981;
          border-bottom: none;
          padding-bottom: 0;
        }

        .legal-section h2 {
          font-size: 1.5rem;
          margin-bottom: var(--space-lg);
          padding-bottom: var(--space-sm);
          border-bottom: 1px solid var(--color-bg-tertiary);
        }

        .legal-section p {
          margin-bottom: var(--space-md);
          max-width: none;
        }

        .legal-section ul {
          list-style: none;
          padding: 0;
          margin: 0 0 var(--space-lg);
        }

        .legal-section li {
          padding: var(--space-sm) 0;
          padding-left: var(--space-lg);
          position: relative;
          color: var(--color-text-secondary);
        }

        .legal-section li::before {
          content: '—';
          position: absolute;
          left: 0;
          color: var(--color-accent);
        }

        .legal-section a {
          color: var(--color-accent);
        }

        .cookie-list {
          display: flex;
          flex-direction: column;
          gap: var(--space-sm);
        }

        .cookie-item {
          display: flex;
          align-items: flex-start;
          gap: var(--space-md);
          padding: var(--space-md) var(--space-lg);
          background: var(--color-bg-secondary);
          border-radius: 8px;
        }

        .cookie-item.no svg {
          color: #ef4444;
          flex-shrink: 0;
          margin-top: 2px;
        }

        .cookie-item.yes svg {
          color: #10b981;
          flex-shrink: 0;
          margin-top: 2px;
        }

        .cookie-item div {
          display: flex;
          flex-direction: column;
        }

        .cookie-item strong {
          font-size: 0.95rem;
          margin-bottom: 2px;
        }

        .cookie-item span {
          font-size: 0.85rem;
          color: var(--color-text-muted);
        }

        .cookie-table {
          width: 100%;
          border-collapse: collapse;
        }

        .cookie-table th,
        .cookie-table td {
          padding: var(--space-md);
          text-align: left;
          border-bottom: 1px solid var(--color-bg-tertiary);
        }

        .cookie-table th {
          font-size: 0.75rem;
          text-transform: uppercase;
          letter-spacing: 0.1em;
          color: var(--color-text-muted);
          background: var(--color-bg-secondary);
        }

        .cookie-table td {
          font-size: 0.9rem;
        }

        .cookie-table code {
          font-family: var(--font-mono);
          font-size: 0.85rem;
          color: var(--color-accent);
        }
      `}</style>
    </div>
  );
}
