// @a0n/gnosis/forkjoin-app/client/pages/DashPage.tsx
/**
 * DashPage - Dedicated page for Dash storage layer
 *
 * Multi-layer edge storage with KV, D1, and R2.
 */

import {
  ArrowLeft,
  Database,
  HardDrive,
  Archive,
  Layers,
  Github,
  BookOpen,
  Code,
} from '@a0n/aeon-icons';
import { getFoundationDocsTrackUrl } from '../docsLinks';

/**
 * Renders the Dash Page view.
 */
export function DashPage() {
  return (
    <div className="product-page dash-page">
      <nav className="product-nav">
        <a href="/" className="back-link">
          <ArrowLeft size={20} />
          <span>Back to Aeon Foundation</span>
        </a>
      </nav>

      <main className="product-content">
        <header className="product-header">
          <Database size={64} className="product-icon" />
          <span className="product-badge">Storage Layer</span>
          <h1>Dash</h1>
          <p className="product-tagline">
            Multi-layer edge storage. 1ms KV. 5ms D1. 10ms R2. No origin server.
            No cold starts.
          </p>
          <div className="product-actions">
            <a
              href="https://github.com/forkjoin-ai/dash"
              className="button button-primary"
              target="_blank"
              rel="noopener noreferrer"
            >
              <Github size={18} />
              View on GitHub
            </a>
            <a
              href={getFoundationDocsTrackUrl('dash')}
              className="button button-secondary"
              target="_blank"
              rel="noopener noreferrer"
            >
              <BookOpen size={18} />
              Documentation
            </a>
          </div>
        </header>

        <section className="product-section">
          <h2>Installation</h2>
          <pre className="install-code">npm install @affectively/dash</pre>
        </section>

        <section className="product-section">
          <h2>Storage Layers</h2>
          <div className="layers-grid">
            <div className="layer-card">
              <div className="layer-header">
                <Database size={32} className="layer-icon" />
                <div className="layer-timing">
                  <span className="timing-value">1ms</span>
                  <span className="timing-label">p50</span>
                </div>
              </div>
              <h3>KV Cache</h3>
              <p>
                Global key-value storage with sub-millisecond reads. Perfect for
                session data, feature flags, and hot paths.
              </p>
              <div className="layer-features">
                <span className="layer-feature">Global replication</span>
                <span className="layer-feature">TTL support</span>
                <span className="layer-feature">Atomic ops</span>
              </div>
            </div>

            <div className="layer-card">
              <div className="layer-header">
                <HardDrive size={32} className="layer-icon" />
                <div className="layer-timing">
                  <span className="timing-value">5ms</span>
                  <span className="timing-label">p50</span>
                </div>
              </div>
              <h3>D1 Database</h3>
              <p>
                SQLite at the edge. Full SQL queries with zero cold starts.
                Migrations, transactions, and joins.
              </p>
              <div className="layer-features">
                <span className="layer-feature">SQLite compatible</span>
                <span className="layer-feature">Transactions</span>
                <span className="layer-feature">Zero cold start</span>
              </div>
            </div>

            <div className="layer-card">
              <div className="layer-header">
                <Archive size={32} className="layer-icon" />
                <div className="layer-timing">
                  <span className="timing-value">10ms</span>
                  <span className="timing-label">p50</span>
                </div>
              </div>
              <h3>R2 Objects</h3>
              <p>
                S3-compatible object storage at the edge. No egress fees. Global
                distribution. Unlimited scale.
              </p>
              <div className="layer-features">
                <span className="layer-feature">S3 compatible</span>
                <span className="layer-feature">No egress fees</span>
                <span className="layer-feature">Infinite scale</span>
              </div>
            </div>

            <div className="layer-card">
              <div className="layer-header">
                <Layers size={32} className="layer-icon" />
                <div className="layer-timing">
                  <span className="timing-value">Auto</span>
                  <span className="timing-label">tiering</span>
                </div>
              </div>
              <h3>Smart Tiering</h3>
              <p>
                Dash automatically tiers data across layers. Hot data in KV.
                Warm data in D1. Cold data in R2.
              </p>
              <div className="layer-features">
                <span className="layer-feature">Access patterns</span>
                <span className="layer-feature">Cost optimized</span>
                <span className="layer-feature">Zero config</span>
              </div>
            </div>
          </div>
        </section>

        <section className="product-section">
          <h2>Storage Hierarchy</h2>
          <div className="hierarchy-diagram">
            <div className="hierarchy-level l1">
              <span className="level-name">L1 Cache</span>
              <span className="level-desc">Worker Memory</span>
              <span className="level-time">~0.1ms</span>
            </div>
            <div className="hierarchy-arrow">↓</div>
            <div className="hierarchy-level l2">
              <span className="level-name">L2 Cache</span>
              <span className="level-desc">KV Store</span>
              <span className="level-time">~1ms</span>
            </div>
            <div className="hierarchy-arrow">↓</div>
            <div className="hierarchy-level l3">
              <span className="level-name">L3 Store</span>
              <span className="level-desc">D1 Database</span>
              <span className="level-time">~5ms</span>
            </div>
            <div className="hierarchy-arrow">↓</div>
            <div className="hierarchy-level l4">
              <span className="level-name">L4 Archive</span>
              <span className="level-desc">R2 Objects</span>
              <span className="level-time">~10ms</span>
            </div>
          </div>
        </section>

        <section className="product-section">
          <h2>Quick Start</h2>
          <div className="code-example">
            <div className="code-header">
              <Code size={18} />
              <span>Basic Usage</span>
            </div>
            <pre className="code-block">{`import { Dash } from '@affectively/dash';

// Connect to edge storage
const dash = new Dash({
  kv: env.CACHE,
  d1: env.DB,
  r2: env.ASSETS,
});

// Read with automatic cache hierarchy
const user = await dash.get('user:123');

// Write with smart tiering
await dash.set('user:123', userData, {
  ttl: 3600,           // KV cache for 1 hour
  persist: true,       // Also write to D1
});

// Query with SQL
const results = await dash.query(
  'SELECT * FROM sessions WHERE user_id = ?',
  [userId]
);

// Store assets
await dash.store('avatar.png', imageData);`}</pre>
          </div>
        </section>

        <section className="product-section">
          <h2>Advanced Usage</h2>
          <div className="code-example">
            <div className="code-header">
              <Code size={18} />
              <span>Smart Tiering Configuration</span>
            </div>
            <pre className="code-block">{`import { Dash, TieringPolicy } from '@affectively/dash';

const dash = new Dash({
  kv: env.CACHE,
  d1: env.DB,
  r2: env.ASSETS,

  tiering: {
    // Define access patterns
    policies: [
      {
        pattern: 'session:*',
        tier: TieringPolicy.HOT,      // Always in KV
        ttl: 3600,
      },
      {
        pattern: 'user:*',
        tier: TieringPolicy.WARM,     // KV + D1
        ttl: 86400,
      },
      {
        pattern: 'archive:*',
        tier: TieringPolicy.COLD,     // D1 + R2
        ttl: null,                     // Never expires
      },
    ],

    // Auto-promote frequently accessed cold data
    promotion: {
      threshold: 10,  // Access count
      window: 3600,   // Within 1 hour
    },

    // Auto-demote stale hot data
    demotion: {
      idleTime: 86400, // 24 hours without access
    },
  },
});`}</pre>
          </div>

          <div className="code-example">
            <div className="code-header">
              <Code size={18} />
              <span>Transactions</span>
            </div>
            <pre className="code-block">{`import { Dash } from '@affectively/dash';

const dash = new Dash({ d1: env.DB });

// Atomic transactions
await dash.transaction(async (tx) => {
  const account = await tx.get('account:123');

  if (account.balance < amount) {
    throw new Error('Insufficient funds');
  }

  await tx.set('account:123', {
    ...account,
    balance: account.balance - amount,
  });

  await tx.set('account:456', {
    ...recipient,
    balance: recipient.balance + amount,
  });

  await tx.insert('transactions', {
    from: '123',
    to: '456',
    amount,
    timestamp: Date.now(),
  });
});`}</pre>
          </div>

          <div className="code-example">
            <div className="code-header">
              <Code size={18} />
              <span>Streaming Large Objects</span>
            </div>
            <pre className="code-block">{`import { Dash } from '@affectively/dash';

const dash = new Dash({ r2: env.ASSETS });

// Upload large files with streaming
await dash.upload('video.mp4', request.body, {
  contentType: 'video/mp4',
  metadata: {
    uploadedBy: userId,
    duration: 120,
  },
});

// Stream to client
const stream = await dash.stream('video.mp4');
return new Response(stream, {
  headers: {
    'Content-Type': 'video/mp4',
    'Accept-Ranges': 'bytes',
  },
});

// Presigned URLs for direct access
const url = await dash.presign('video.mp4', {
  expiresIn: 3600,
  method: 'GET',
});`}</pre>
          </div>
        </section>

        <section className="product-section">
          <h2>Resources</h2>
          <div className="resources-grid">
            <a
              href="https://github.com/forkjoin-ai/dash"
              className="resource-card"
              target="_blank"
              rel="noopener noreferrer"
            >
              <Github size={24} />
              <h4>GitHub</h4>
              <p>Source code, issues, and discussions</p>
            </a>
            <a
              href={getFoundationDocsTrackUrl('dash')}
              className="resource-card"
              target="_blank"
              rel="noopener noreferrer"
            >
              <BookOpen size={24} />
              <h4>Documentation</h4>
              <p>Full API reference and guides</p>
            </a>
            <a
              href="https://github.com/forkjoin-ai/dash/tree/main/examples"
              className="resource-card"
              target="_blank"
              rel="noopener noreferrer"
            >
              <Code size={24} />
              <h4>Examples</h4>
              <p>Sample projects and use cases</p>
            </a>
          </div>
        </section>
      </main>

      <style>{`
        .dash-page {
          --product-color: #f59e0b;
          min-height: 100vh;
          background: var(--color-bg);
        }

        .product-nav {
          padding: var(--space-lg) var(--space-xl);
          border-bottom: 1px solid var(--color-bg-tertiary);
        }

        .back-link {
          display: inline-flex;
          align-items: center;
          gap: var(--space-sm);
          color: var(--color-text-secondary);
          text-decoration: none;
          transition: color 0.2s ease;
        }

        .back-link:hover {
          color: var(--color-text);
        }

        .product-content {
          max-width: 900px;
          margin: 0 auto;
          padding: var(--space-3xl) var(--space-xl);
        }

        .product-header {
          text-align: center;
          margin-bottom: var(--space-3xl);
        }

        .product-icon {
          display: block;
          margin: 0 auto var(--space-lg);
          color: var(--product-color);
        }

        .product-badge {
          display: inline-block;
          padding: 0.25rem 0.75rem;
          font-size: 0.75rem;
          font-weight: 500;
          text-transform: uppercase;
          letter-spacing: 0.1em;
          color: var(--product-color);
          border: 1px solid var(--product-color);
          border-radius: var(--radius-full);
          margin-bottom: var(--space-md);
        }

        .product-header h1 {
          font-size: 3rem;
          margin-bottom: var(--space-md);
        }

        .product-tagline {
          font-size: 1.25rem;
          color: var(--color-text-secondary);
          max-width: 600px;
          margin: 0 auto var(--space-xl);
        }

        .product-actions {
          display: flex;
          gap: var(--space-md);
          justify-content: center;
          flex-wrap: wrap;
        }

        .product-actions .button {
          display: inline-flex;
          align-items: center;
          gap: var(--space-sm);
        }

        .product-section {
          margin-bottom: var(--space-3xl);
        }

        .product-section h2 {
          font-size: 1.5rem;
          margin-bottom: var(--space-lg);
          padding-bottom: var(--space-sm);
          border-bottom: 1px solid var(--color-bg-tertiary);
        }

        .install-code {
          font-family: var(--font-mono);
          font-size: 1.1rem;
          padding: var(--space-lg) var(--space-xl);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
          color: var(--product-color);
        }

        .layers-grid {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(260px, 1fr));
          gap: var(--space-lg);
        }

        .layer-card {
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 16px;
          padding: var(--space-xl);
        }

        .layer-header {
          display: flex;
          justify-content: space-between;
          align-items: flex-start;
          margin-bottom: var(--space-md);
        }

        .layer-icon {
          color: var(--product-color);
        }

        .layer-timing {
          display: flex;
          flex-direction: column;
          align-items: flex-end;
        }

        .timing-value {
          font-family: var(--font-mono);
          font-size: 1.25rem;
          font-weight: 600;
          color: #10b981;
        }

        .timing-label {
          font-size: 0.65rem;
          text-transform: uppercase;
          letter-spacing: 0.1em;
          color: var(--color-text-muted);
        }

        .layer-card h3 {
          margin-bottom: var(--space-sm);
          color: var(--product-color);
        }

        .layer-card p {
          color: var(--color-text-secondary);
          font-size: 0.9rem;
          margin-bottom: var(--space-md);
        }

        .layer-features {
          display: flex;
          flex-wrap: wrap;
          gap: var(--space-xs);
        }

        .layer-feature {
          font-size: 0.7rem;
          padding: 4px 8px;
          background: rgba(245, 158, 11, 0.1);
          border: 1px solid rgba(245, 158, 11, 0.2);
          border-radius: 4px;
          color: var(--color-text-muted);
        }

        .hierarchy-diagram {
          display: flex;
          flex-direction: column;
          align-items: center;
          gap: var(--space-xs);
          max-width: 500px;
          margin: 0 auto;
        }

        .hierarchy-level {
          display: flex;
          justify-content: space-between;
          align-items: center;
          width: 100%;
          padding: var(--space-md) var(--space-lg);
          border-radius: 8px;
        }

        .hierarchy-level.l1 {
          background: rgba(16, 185, 129, 0.15);
          border: 1px solid rgba(16, 185, 129, 0.3);
        }

        .hierarchy-level.l2 {
          background: rgba(6, 182, 212, 0.15);
          border: 1px solid rgba(6, 182, 212, 0.3);
        }

        .hierarchy-level.l3 {
          background: rgba(51, 102, 153, 0.15);
          border: 1px solid rgba(51, 102, 153, 0.3);
        }

        .hierarchy-level.l4 {
          background: rgba(245, 158, 11, 0.15);
          border: 1px solid rgba(245, 158, 11, 0.3);
        }

        .level-name {
          font-weight: 600;
          font-size: 0.85rem;
          color: var(--color-text);
        }

        .level-desc {
          font-size: 0.8rem;
          color: var(--color-text-secondary);
        }

        .level-time {
          font-family: var(--font-mono);
          font-size: 0.85rem;
          color: #10b981;
        }

        .hierarchy-arrow {
          color: var(--color-text-muted);
          font-size: 0.85rem;
        }

        .code-example {
          margin-bottom: var(--space-xl);
        }

        .code-header {
          display: flex;
          align-items: center;
          gap: var(--space-sm);
          margin-bottom: var(--space-sm);
          color: var(--color-text-secondary);
          font-size: 0.9rem;
        }

        .code-header svg {
          color: var(--product-color);
        }

        .code-block {
          font-family: var(--font-mono);
          font-size: 0.85rem;
          line-height: 1.6;
          padding: var(--space-lg);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
          overflow-x: auto;
          margin: 0;
        }

        .resources-grid {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(250px, 1fr));
          gap: var(--space-lg);
        }

        .resource-card {
          padding: var(--space-xl);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
          text-decoration: none;
          color: inherit;
          transition: all 0.2s ease;
        }

        .resource-card:hover {
          border-color: var(--product-color);
          transform: translateY(-2px);
        }

        .resource-card svg {
          color: var(--product-color);
          margin-bottom: var(--space-md);
        }

        .resource-card h4 {
          margin-bottom: var(--space-xs);
        }

        .resource-card p {
          color: var(--color-text-secondary);
          font-size: 0.85rem;
          margin: 0;
        }
      `}</style>
    </div>
  );
}
