// @a0n/gnosis/forkjoin-app/client/pages/DashRelayPage.tsx
/**
 * DashRelayPage - Dedicated page for the dashrelay platform
 *
 * Crypto relay infrastructure for local-first apps.
 * Messaging + CRDT brokering with UCAN identity and zero-knowledge encryption.
 */

import {
  ArrowLeft,
  Zap,
  Fingerprint,
  Lock,
  Shield,
  Globe,
  Github,
  BookOpen,
  Code,
} from '@a0n/aeon-icons';
import {
  DASHRELAY_TRUST_LOG_DESCRIPTION,
  TERM_CRYPTO_RELAY,
  TERM_DASHRELAY,
} from '../../src/terminology';
import { getFoundationDocsTrackUrl } from '../docsLinks';

/**
 * Renders the Dash Relay Page view.
 */
export function DashRelayPage() {
  return (
    <div className="product-page dashrelay-page">
      <nav className="product-nav">
        <a href="/" className="back-link">
          <ArrowLeft size={20} />
          <span>Back to Aeon Foundation</span>
        </a>
      </nav>

      <main className="product-content">
        <header className="product-header">
          <Zap size={64} className="product-icon" />
          <span className="product-badge">{TERM_CRYPTO_RELAY}</span>
          <h1>{TERM_DASHRELAY}</h1>
          <p className="product-tagline">{DASHRELAY_TRUST_LOG_DESCRIPTION}</p>
          <div className="product-actions">
            <a
              href="https://github.com/forkjoin-ai/dashrelay"
              className="button button-primary"
              target="_blank"
              rel="noopener noreferrer"
            >
              <Github size={18} />
              View on GitHub
            </a>
            <a
              href={getFoundationDocsTrackUrl('dashrelay')}
              className="button button-secondary"
              target="_blank"
              rel="noopener noreferrer"
            >
              <BookOpen size={18} />
              Documentation
            </a>
          </div>
        </header>

        <section className="product-section">
          <h2>Installation</h2>
          <pre className="install-code">npm install @dashrelay/client</pre>
        </section>

        <section className="product-section">
          <h2>Core Capabilities</h2>
          <div className="features-grid">
            <div className="feature-card">
              <Fingerprint size={40} className="feature-icon" />
              <h3>UCAN Identity</h3>
              <p>
                Decentralized identity with did:key DIDs and ECDSA P-256. Users
                own their keys. No central authority controls access.
                Capability-based delegation chains let you grant scoped
                permissions without a server.
              </p>
              <div className="capability-chain">
                <span className="cap">did:key:z6Mk...</span>
                <span className="cap-arrow" aria-hidden="true">
                  &rarr;
                </span>
                <span className="cap">room:write</span>
                <span className="cap-arrow" aria-hidden="true">
                  &rarr;
                </span>
                <span className="cap">delegated</span>
              </div>
            </div>

            <div className="feature-card">
              <Lock size={40} className="feature-icon zk" />
              <h3>Zero-Knowledge Encryption</h3>
              <p>
                ECIES-P256 key exchange with AES-256-GCM payload encryption. The
                relay server never sees plaintext. Forward secrecy via ephemeral
                ECDH means compromising one session does not expose others.
              </p>
              <div className="encryption-flow">
                <span className="enc-step">Plaintext</span>
                <span className="enc-arrow" aria-hidden="true">
                  &rarr;
                </span>
                <span className="enc-step enc-encrypted">ECIES + AES-GCM</span>
                <span className="enc-arrow" aria-hidden="true">
                  &rarr;
                </span>
                <span className="enc-step">Ciphertext relay</span>
                <span className="enc-arrow" aria-hidden="true">
                  &rarr;
                </span>
                <span className="enc-step">Plaintext</span>
              </div>
            </div>

            <div className="feature-card">
              <Shield size={40} className="feature-icon" />
              <h3>Messaging + CRDT Broker</h3>
              <p>
                Built on Cloudflare Durable Objects with WebSocket hibernation.
                Room-scoped messages and CRDT operations are relayed as
                ciphertext and anchored in a verifiable trust log. Rooms
                hibernate when empty for zero idle cost.
              </p>
              <div className="stats-row">
                <div className="stat">
                  <span className="stat-value">0</span>
                  <span className="stat-label">plaintext</span>
                </div>
                <div className="stat">
                  <span className="stat-value">$0</span>
                  <span className="stat-label">idle cost</span>
                </div>
                <div className="stat">
                  <span className="stat-value">E2E</span>
                  <span className="stat-label">encrypted</span>
                </div>
              </div>
            </div>

            <div className="feature-card">
              <Globe size={40} className="feature-icon" />
              <h3>Global Edge Network</h3>
              <p>
                300+ points of presence via Cloudflare. Sub-50ms global message
                fanout. Regional sharding keeps data close to users. V8 isolates
                handle auth validation at the edge with no origin round-trip.
              </p>
              <div className="stats-row">
                <div className="stat">
                  <span className="stat-value">300+</span>
                  <span className="stat-label">PoPs</span>
                </div>
                <div className="stat">
                  <span className="stat-value">&lt;50ms</span>
                  <span className="stat-label">global</span>
                </div>
                <div className="stat">
                  <span className="stat-value">0</span>
                  <span className="stat-label">origin</span>
                </div>
              </div>
            </div>
          </div>
        </section>

        <section className="product-section">
          <h2>Quick Start</h2>
          <div className="code-example">
            <div className="code-header">
              <Code size={18} />
              <span>Yjs Provider</span>
            </div>
            <pre className="code-block">{`import { DashRelay } from '@dashrelay/client';
import { QDoc } from '@a0n/gnosis';

const keypair = await DashRelay.generateKeypair();
const ucan = await DashRelay.createUcan(keypair);

const doc = new QDoc();
const relay = new DashRelay({
  ucan,
  roomName: 'my-document',
  encrypt: true,
});

relay.connect(doc);
// Encrypted, authenticated, syncing globally.`}</pre>
          </div>

          <div className="code-example">
            <div className="code-header">
              <Code size={18} />
              <span>Automerge Provider</span>
            </div>
            <pre className="code-block">{`import { DashRelay } from '@dashrelay/client';
import * as Automerge from '@automerge/automerge';

const keypair = await DashRelay.generateKeypair();
const ucan = await DashRelay.createUcan(keypair);

let doc = Automerge.init();
const relay = new DashRelay({
  ucan,
  roomName: 'my-document',
  encrypt: true,
});

relay.connectAutomerge(doc, {
  onChange: (newDoc) => { doc = newDoc; },
});
// Same relay, same encryption, different CRDT.`}</pre>
          </div>
        </section>

        <section className="product-section">
          <h2>Advanced Usage</h2>
          <div className="code-example">
            <div className="code-header">
              <Code size={18} />
              <span>Capability Delegation</span>
            </div>
            <pre className="code-block">{`import { DashRelay, UcanBuilder } from '@dashrelay/client';

// Create a scoped capability for a collaborator
const delegation = await UcanBuilder.create()
  .issuedBy(myKeypair)
  .forAudience(collaboratorDid)
  .withCapability({
    resource: 'room:my-document',
    ability: 'write',
    caveats: {
      expiresAt: Date.now() + 3600_000, // 1 hour
    },
  })
  .build();

// Collaborator connects with delegated capability
const relay = new DashRelay({
  ucan: delegation,
  roomName: 'my-document',
  encrypt: true,
});`}</pre>
          </div>

          <div className="code-example">
            <div className="code-header">
              <Code size={18} />
              <span>Presence and Awareness</span>
            </div>
            <pre className="code-block">{`import { DashRelay } from '@dashrelay/client';
import { QDoc } from '@a0n/gnosis';

const doc = new QDoc();
const relay = new DashRelay({
  ucan,
  roomName: 'my-document',
  encrypt: true,
});

relay.connect(doc);

// Built-in awareness for presence
const awareness = relay.awareness;

awareness.setLocalState({
  user: { name: 'Alice', color: '#10b981' },
  cursor: { x: 100, y: 200 },
});

// Listen for other users
awareness.on('change', () => {
  const states = awareness.getStates();
  for (const [clientId, state] of states) {
    renderCursor(clientId, state);
  }
});`}</pre>
          </div>

          <div className="code-example">
            <div className="code-header">
              <Code size={18} />
              <span>Offline-First with Auto-Sync</span>
            </div>
            <pre className="code-block">{`import { DashRelay } from '@dashrelay/client';
import { QDoc } from '@a0n/gnosis';

const doc = new QDoc();
const relay = new DashRelay({
  ucan,
  roomName: 'my-document',
  encrypt: true,
  offline: {
    persist: true,     // Persist to IndexedDB
    autoSync: true,    // Auto-sync on reconnect
  },
});

relay.connect(doc);

// Works offline - changes queue locally
doc.getMap('data').set('key', 'value');

// Monitor connection state
relay.on('status', ({ status }) => {
  if (status === 'connected') {
    console.log('Online - syncing queued changes');
  } else {
    console.log('Offline - changes queued locally');
  }
});`}</pre>
          </div>
        </section>

        <section className="product-section">
          <h2>Built On</h2>
          <p className="section-intro">
            dashrelay combines the best of the Aeon Foundation stack into a
            managed platform.
          </p>
          <div className="built-on-grid">
            <a href="#/relay" className="built-on-card">
              <h4>Relay</h4>
              <p>WebTransport mesh networking</p>
            </a>
            <a href="#/dash" className="built-on-card">
              <h4>Dash</h4>
              <p>Multi-layer edge storage</p>
            </a>
            <a href="#/aegis" className="built-on-card">
              <h4>Aegis</h4>
              <p>UCAN identity and ZK proofs</p>
            </a>
            <a href="#/aeon" className="built-on-card">
              <h4>Aeon</h4>
              <p>CRDT synchronization</p>
            </a>
          </div>
        </section>

        <section className="product-section">
          <h2>Resources</h2>
          <div className="resources-grid">
            <a
              href="https://github.com/forkjoin-ai/dashrelay"
              className="resource-card"
              target="_blank"
              rel="noopener noreferrer"
            >
              <Github size={24} />
              <h4>GitHub</h4>
              <p>Source code, issues, and discussions</p>
            </a>
            <a
              href={getFoundationDocsTrackUrl('dashrelay')}
              className="resource-card"
              target="_blank"
              rel="noopener noreferrer"
            >
              <BookOpen size={24} />
              <h4>Documentation</h4>
              <p>Full API reference and guides</p>
            </a>
            <a
              href="https://dashrelay.com"
              className="resource-card"
              target="_blank"
              rel="noopener noreferrer"
            >
              <Zap size={24} />
              <h4>dashrelay.com</h4>
              <p>Dashboard, API keys, and pricing</p>
            </a>
          </div>
        </section>
      </main>

      <style>{`
        .dashrelay-page {
          --product-color: #10b981;
          min-height: 100vh;
          background: var(--color-bg);
        }

        .product-nav {
          padding: var(--space-lg) var(--space-xl);
          border-bottom: 1px solid var(--color-bg-tertiary);
        }

        .back-link {
          display: inline-flex;
          align-items: center;
          gap: var(--space-sm);
          color: var(--color-text-secondary);
          text-decoration: none;
          transition: color 0.2s ease;
        }

        .back-link:hover {
          color: var(--color-text);
        }

        .product-content {
          max-width: 900px;
          margin: 0 auto;
          padding: var(--space-3xl) var(--space-xl);
        }

        .product-header {
          text-align: center;
          margin-bottom: var(--space-3xl);
        }

        .product-icon {
          display: block;
          margin: 0 auto var(--space-lg);
          color: var(--product-color);
        }

        .product-badge {
          display: inline-block;
          padding: 0.25rem 0.75rem;
          font-size: 0.75rem;
          font-weight: 500;
          text-transform: uppercase;
          letter-spacing: 0.1em;
          color: var(--product-color);
          border: 1px solid var(--product-color);
          border-radius: var(--radius-full);
          margin-bottom: var(--space-md);
        }

        .product-header h1 {
          font-size: 3rem;
          margin-bottom: var(--space-md);
        }

        .product-tagline {
          font-size: 1.25rem;
          color: var(--color-text-secondary);
          max-width: 600px;
          margin: 0 auto var(--space-xl);
        }

        .product-actions {
          display: flex;
          gap: var(--space-md);
          justify-content: center;
          flex-wrap: wrap;
        }

        .product-actions .button {
          display: inline-flex;
          align-items: center;
          gap: var(--space-sm);
        }

        .product-section {
          margin-bottom: var(--space-3xl);
        }

        .product-section h2 {
          font-size: 1.5rem;
          margin-bottom: var(--space-lg);
          padding-bottom: var(--space-sm);
          border-bottom: 1px solid var(--color-bg-tertiary);
        }

        .install-code {
          font-family: var(--font-mono);
          font-size: 1.1rem;
          padding: var(--space-lg) var(--space-xl);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
          color: var(--product-color);
        }

        .features-grid {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(280px, 1fr));
          gap: var(--space-xl);
        }

        .feature-card {
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 16px;
          padding: var(--space-xl);
        }

        .feature-icon {
          color: var(--product-color);
          margin-bottom: var(--space-md);
        }

        .feature-icon.zk {
          color: #8b5cf6;
        }

        .feature-card h3 {
          margin-bottom: var(--space-sm);
          color: var(--product-color);
        }

        .feature-card p {
          color: var(--color-text-secondary);
          margin-bottom: var(--space-lg);
        }

        .capability-chain {
          display: flex;
          align-items: center;
          gap: var(--space-xs);
          flex-wrap: wrap;
        }

        .cap {
          font-family: var(--font-mono);
          font-size: 0.75rem;
          padding: 4px 8px;
          background: rgba(16, 185, 129, 0.1);
          border: 1px solid rgba(16, 185, 129, 0.3);
          border-radius: 4px;
          color: var(--product-color);
        }

        .cap-arrow {
          color: var(--color-text-muted);
          font-size: 0.85rem;
        }

        .encryption-flow {
          display: flex;
          align-items: center;
          gap: var(--space-xs);
          flex-wrap: wrap;
        }

        .enc-step {
          font-size: 0.75rem;
          padding: 4px 8px;
          background: rgba(16, 185, 129, 0.1);
          border: 1px solid rgba(16, 185, 129, 0.2);
          border-radius: 4px;
          color: var(--color-text-secondary);
        }

        .enc-step.enc-encrypted {
          background: rgba(139, 92, 246, 0.15);
          border-color: rgba(139, 92, 246, 0.3);
          color: #8b5cf6;
        }

        .enc-arrow {
          color: var(--color-text-muted);
          font-size: 0.85rem;
        }

        .stats-row {
          display: flex;
          gap: var(--space-lg);
        }

        .stat {
          display: flex;
          flex-direction: column;
          align-items: center;
        }

        .stat-value {
          font-family: var(--font-mono);
          font-size: 1.5rem;
          font-weight: 600;
          color: var(--product-color);
        }

        .stat-label {
          font-size: 0.7rem;
          text-transform: uppercase;
          letter-spacing: 0.1em;
          color: var(--color-text-muted);
        }

        .section-intro {
          color: var(--color-text-secondary);
          margin-bottom: var(--space-lg);
        }

        .built-on-grid {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(200px, 1fr));
          gap: var(--space-md);
        }

        .built-on-card {
          padding: var(--space-lg);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
          text-decoration: none;
          color: inherit;
          transition: all 0.2s ease;
        }

        .built-on-card:hover {
          border-color: var(--product-color);
          transform: translateY(-2px);
        }

        .built-on-card h4 {
          color: var(--product-color);
          margin-bottom: var(--space-xs);
        }

        .built-on-card p {
          color: var(--color-text-secondary);
          font-size: 0.85rem;
          margin: 0;
        }

        .code-example {
          margin-bottom: var(--space-xl);
        }

        .code-header {
          display: flex;
          align-items: center;
          gap: var(--space-sm);
          margin-bottom: var(--space-sm);
          color: var(--color-text-secondary);
          font-size: 0.9rem;
        }

        .code-header svg {
          color: var(--product-color);
        }

        .code-block {
          font-family: var(--font-mono);
          font-size: 0.85rem;
          line-height: 1.6;
          padding: var(--space-lg);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
          overflow-x: auto;
          margin: 0;
        }

        .resources-grid {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(250px, 1fr));
          gap: var(--space-lg);
        }

        .resource-card {
          padding: var(--space-xl);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
          text-decoration: none;
          color: inherit;
          transition: all 0.2s ease;
        }

        .resource-card:hover {
          border-color: var(--product-color);
          transform: translateY(-2px);
        }

        .resource-card svg {
          color: var(--product-color);
          margin-bottom: var(--space-md);
        }

        .resource-card h4 {
          margin-bottom: var(--space-xs);
        }

        .resource-card p {
          color: var(--color-text-secondary);
          font-size: 0.85rem;
          margin: 0;
        }
      `}</style>
    </div>
  );
}
