// @a0n/gnosis/forkjoin-app/client/pages/EdgeworkPage.tsx
/**
 * EdgeworkPage - Dedicated page for Edgework SDK
 *
 * Client-side AI, real-time sync, on-device training.
 * Zero server dependency.
 */

import {
  ArrowLeft,
  Cpu,
  RefreshCcw,
  Lock,
  Zap,
  Lightbulb,
  Database,
  Github,
  BookOpen,
  Code,
  ExternalLink,
} from '@a0n/aeon-icons';
import { getFoundationDocsTrackUrl } from '../docsLinks';

/**
 * Renders the Edgework Page view.
 */
export function EdgeworkPage() {
  return (
    <div className="product-page edgework-page">
      <nav className="product-nav">
        <a href="/" className="back-link">
          <ArrowLeft size={20} />
          <span>Back to Aeon Foundation</span>
        </a>
      </nav>

      <main className="product-content">
        <header className="product-header">
          <Zap size={64} className="product-icon" />
          <span className="product-badge">Client SDK</span>
          <h1>Edgework SDK</h1>
          <p className="product-tagline">
            Client-side AI. Real-time sync. On-device training.
            <br />
            <strong>Zero server dependency.</strong>
          </p>
          <div className="product-actions">
            <a
              href="https://github.com/forkjoin-ai/edgework-sdk"
              className="button button-primary"
              target="_blank"
              rel="noopener noreferrer"
            >
              <Github size={18} />
              View on GitHub
            </a>
            <a
              href="https://npmjs.com/package/@affectively/edgework-sdk"
              className="button button-secondary"
              target="_blank"
              rel="noopener noreferrer"
            >
              <ExternalLink size={18} />
              npm
            </a>
          </div>
        </header>

        <section className="product-section">
          <h2>Installation</h2>
          <pre className="install-code">
            npm install @affectively/edgework-sdk
          </pre>
        </section>

        <section className="product-section">
          <h2>The Vision</h2>
          <p className="vision-text">
            We didn't just build an SDK. We built the infrastructure for a new
            kind of internet—one where your data never leaves your device unless
            you want it to.
          </p>
        </section>

        <section className="product-section">
          <h2>What You Get</h2>
          <div className="features-grid">
            <div className="feature-card">
              <Cpu size={32} className="feature-icon" />
              <h4>AI That Runs on Your Device</h4>
              <p>
                No waiting for servers. AI runs directly in your browser, giving
                you instant responses that never leave your device.
              </p>
            </div>
            <div className="feature-card">
              <RefreshCcw size={32} className="feature-icon" />
              <h4>Work Offline, Sync When Ready</h4>
              <p>
                Edit anywhere, anytime—even without internet. Your changes sync
                automatically when you reconnect. Never lose work.
              </p>
            </div>
            <div className="feature-card">
              <Lock size={32} className="feature-icon" />
              <h4>You Control Your Data</h4>
              <p>
                No central authority decides who has access. You grant
                permissions, and you can revoke them anytime.
              </p>
            </div>
            <div className="feature-card">
              <Zap size={32} className="feature-icon" />
              <h4>Connect Your Way</h4>
              <p>
                Whether you're syncing between devices or collaborating with
                others, choose the connection that works best for your
                situation.
              </p>
            </div>
            <div className="feature-card">
              <Lightbulb size={32} className="feature-icon" />
              <h4>A Model That Learns from You</h4>
              <p>
                Your feedback makes your AI better—and only yours. The
                improvements stay on your device, private by design.
              </p>
            </div>
            <div className="feature-card">
              <Database size={32} className="feature-icon" />
              <h4>Your Data, Your Device</h4>
              <p>
                Everything lives on your device first. Cloud sync is
                optional—and you decide what gets shared.
              </p>
            </div>
          </div>
        </section>

        <section className="product-section">
          <h2>Statistics</h2>
          <div className="stats-grid">
            <div className="stat-card">
              <span className="stat-value">0ms</span>
              <span className="stat-label">Server Latency</span>
            </div>
            <div className="stat-card">
              <span className="stat-value">100%</span>
              <span className="stat-label">Offline Capable</span>
            </div>
            <div className="stat-card">
              <span className="stat-value">300+</span>
              <span className="stat-label">Edge Locations</span>
            </div>
            <div className="stat-card">
              <span className="stat-value">0</span>
              <span className="stat-label">Data Leakage</span>
            </div>
          </div>
        </section>

        <section className="product-section">
          <h2>The Stack</h2>
          <div className="stack-visual">
            <div className="stack-layer">
              <span className="layer-name">React Hooks</span>
              <span className="layer-desc">
                useInference, useStorage, useFeedback, useSync
              </span>
            </div>
            <div className="stack-layer">
              <span className="layer-name">EdgeworkProvider</span>
              <span className="layer-desc">
                Context that wires everything together
              </span>
            </div>
            <div className="stack-layer">
              <span className="layer-name">Inference Engine</span>
              <span className="layer-desc">
                WebGPU + ONNX runtime for on-device AI
              </span>
            </div>
            <div className="stack-layer">
              <span className="layer-name">Storage Layer</span>
              <span className="layer-desc">
                IndexedDB with CRDT conflict resolution
              </span>
            </div>
            <div className="stack-layer">
              <span className="layer-name">Sync Protocol</span>
              <span className="layer-desc">
                WebSocket/WebTransport/WebRTC hub-and-spoke
              </span>
            </div>
            <div className="stack-layer">
              <span className="layer-name">Auth Layer</span>
              <span className="layer-desc">
                UCAN capability tokens, no central authority
              </span>
            </div>
          </div>
        </section>

        <section className="product-section">
          <h2>Quick Start</h2>
          <div className="code-example">
            <div className="code-header">
              <Code size={18} />
              <span>1. Wrap your app</span>
            </div>
            <pre className="code-block">{`import { EdgeworkProvider } from '@affectively/edgework-sdk/react';

function App({ children }) {
  return (
    <EdgeworkProvider
      config={{
        storage: { dbName: 'my-app', sync: true },
        inference: { model: 'emotion-v1' },
        auth: { ucan: true }
      }}
    >
      {children}
    </EdgeworkProvider>
  );
}`}</pre>
          </div>

          <div className="code-example">
            <div className="code-header">
              <Code size={18} />
              <span>2. Use anywhere</span>
            </div>
            <pre className="code-block">{`import { useInference, useStorage, useSync } from '@affectively/edgework-sdk/react';

function MyComponent() {
  const { infer, isReady } = useInference();
  const { get, set } = useStorage();
  const { status, peers } = useSync();

  // Run AI inference on-device
  const result = await infer({ text: 'How are you feeling?' });

  // Store locally, sync globally
  await set('mood', { emotion: result.primary, timestamp: Date.now() });

  // Check sync status
  console.log(\`Connected to \${peers.length} peers\`);
}`}</pre>
          </div>

          <div className="code-example">
            <div className="code-header">
              <Code size={18} />
              <span>3. Collect feedback (optional RLHF)</span>
            </div>
            <pre className="code-block">{`import { useFeedback } from '@affectively/edgework-sdk/react';

function FeedbackUI({ responseId }) {
  const { submit, pending } = useFeedback();

  return (
    <div>
      <button onClick={() => submit(responseId, 'positive')}>+1</button>
      <button onClick={() => submit(responseId, 'negative')}>-1</button>
    </div>
  );
}
// Feedback trains YOUR local model. Private. Personal. Improving.`}</pre>
          </div>
        </section>

        <section className="product-section">
          <h2>Resources</h2>
          <div className="resources-grid">
            <a
              href="https://github.com/forkjoin-ai/edgework-sdk"
              className="resource-card"
              target="_blank"
              rel="noopener noreferrer"
            >
              <Github size={24} />
              <h4>GitHub</h4>
              <p>Source code, issues, and discussions</p>
            </a>
            <a
              href={getFoundationDocsTrackUrl('edgework')}
              className="resource-card"
              target="_blank"
              rel="noopener noreferrer"
            >
              <BookOpen size={24} />
              <h4>Documentation</h4>
              <p>Full API reference and guides</p>
            </a>
            <a
              href="https://github.com/forkjoin-ai/edgework-sdk/tree/main/examples"
              className="resource-card"
              target="_blank"
              rel="noopener noreferrer"
            >
              <Code size={24} />
              <h4>Examples</h4>
              <p>Sample projects and use cases</p>
            </a>
          </div>
        </section>

        <section className="product-section cta-section">
          <h2>The future is local.</h2>
          <p className="cta-subtitle">Start building it today.</p>
          <div className="product-actions">
            <a
              href="https://npmjs.com/package/@affectively/edgework-sdk"
              className="button button-primary"
              target="_blank"
              rel="noopener noreferrer"
            >
              Install from npm
            </a>
            <a
              href="https://github.com/forkjoin-ai/edgework-sdk"
              className="button button-secondary"
              target="_blank"
              rel="noopener noreferrer"
            >
              View on GitHub
            </a>
          </div>
        </section>
      </main>

      <style>{`
        .edgework-page {
          --product-color: #3b82f6;
          --product-color-light: #60a5fa;
          min-height: 100vh;
          background: #000000;
          color: #f8fafc;
        }

        .product-nav {
          padding: var(--space-lg) var(--space-xl);
          border-bottom: 1px solid rgba(255, 255, 255, 0.1);
        }

        .back-link {
          display: inline-flex;
          align-items: center;
          gap: var(--space-sm);
          color: rgba(248, 250, 252, 0.7);
          text-decoration: none;
          transition: color 0.2s ease;
        }

        .back-link:hover {
          color: #f8fafc;
        }

        .product-content {
          max-width: 900px;
          margin: 0 auto;
          padding: var(--space-3xl) var(--space-xl);
        }

        .product-header {
          text-align: center;
          margin-bottom: var(--space-3xl);
        }

        .product-icon {
          display: block;
          margin: 0 auto var(--space-lg);
          color: var(--product-color);
        }

        .product-badge {
          display: inline-block;
          padding: 0.25rem 0.75rem;
          font-size: 0.75rem;
          font-weight: 500;
          text-transform: uppercase;
          letter-spacing: 0.1em;
          color: var(--product-color);
          border: 1px solid var(--product-color);
          border-radius: var(--radius-full);
          margin-bottom: var(--space-md);
          background: rgba(59, 130, 246, 0.1);
        }

        .product-header h1 {
          font-size: 3rem;
          margin-bottom: var(--space-md);
          color: #f8fafc;
        }

        .product-tagline {
          font-size: 1.25rem;
          color: rgba(248, 250, 252, 0.7);
          max-width: 600px;
          margin: 0 auto var(--space-xl);
          line-height: 1.6;
        }

        .product-actions {
          display: flex;
          gap: var(--space-md);
          justify-content: center;
          flex-wrap: wrap;
        }

        .product-actions .button {
          display: inline-flex;
          align-items: center;
          gap: var(--space-sm);
        }

        .product-section {
          margin-bottom: var(--space-3xl);
        }

        .product-section h2 {
          font-size: 1.5rem;
          margin-bottom: var(--space-lg);
          padding-bottom: var(--space-sm);
          border-bottom: 1px solid rgba(255, 255, 255, 0.1);
          color: #f8fafc;
        }

        .install-code {
          font-family: var(--font-mono);
          font-size: 1.1rem;
          padding: var(--space-lg) var(--space-xl);
          background: rgba(59, 130, 246, 0.1);
          border: 1px solid rgba(59, 130, 246, 0.2);
          border-radius: 12px;
          color: var(--product-color-light);
        }

        .vision-text {
          font-size: 1.1rem;
          line-height: 1.8;
          color: rgba(248, 250, 252, 0.8);
          max-width: 700px;
        }

        .features-grid {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(280px, 1fr));
          gap: var(--space-lg);
        }

        .feature-card {
          padding: var(--space-xl);
          background: rgba(255, 255, 255, 0.02);
          border: 1px solid rgba(255, 255, 255, 0.1);
          border-radius: 12px;
          transition: all 0.2s ease;
        }

        .feature-card:hover {
          border-color: var(--product-color);
          background: rgba(59, 130, 246, 0.05);
        }

        .feature-icon {
          color: var(--product-color);
          margin-bottom: var(--space-md);
        }

        .feature-card h4 {
          color: #f8fafc;
          margin-bottom: var(--space-sm);
          font-size: 1.1rem;
        }

        .feature-card p {
          color: rgba(248, 250, 252, 0.6);
          font-size: 0.9rem;
          margin: 0;
          line-height: 1.5;
        }

        .stats-grid {
          display: flex;
          justify-content: center;
          gap: var(--space-3xl);
          flex-wrap: wrap;
        }

        .stat-card {
          text-align: center;
        }

        .stat-value {
          display: block;
          font-family: var(--font-mono);
          font-size: 2.5rem;
          font-weight: 600;
          color: var(--product-color);
        }

        .stat-label {
          font-size: 0.8rem;
          color: rgba(248, 250, 252, 0.5);
          text-transform: uppercase;
          letter-spacing: 0.1em;
        }

        .stack-visual {
          display: flex;
          flex-direction: column;
          gap: var(--space-sm);
        }

        .stack-layer {
          display: flex;
          justify-content: space-between;
          align-items: center;
          padding: var(--space-md) var(--space-lg);
          background: rgba(59, 130, 246, 0.05);
          border-left: 3px solid var(--product-color);
          border-radius: 0 8px 8px 0;
        }

        .layer-name {
          font-weight: 600;
          color: #f8fafc;
        }

        .layer-desc {
          color: rgba(248, 250, 252, 0.5);
          font-size: 0.85rem;
        }

        .code-example {
          margin-bottom: var(--space-xl);
        }

        .code-header {
          display: flex;
          align-items: center;
          gap: var(--space-sm);
          margin-bottom: var(--space-sm);
          color: rgba(248, 250, 252, 0.7);
          font-size: 0.9rem;
        }

        .code-header svg {
          color: var(--product-color-light);
        }

        .code-block {
          font-family: var(--font-mono);
          font-size: 0.8rem;
          line-height: 1.6;
          padding: var(--space-lg);
          background: rgba(255, 255, 255, 0.03);
          border: 1px solid rgba(255, 255, 255, 0.1);
          border-radius: 12px;
          overflow-x: auto;
          margin: 0;
          color: rgba(248, 250, 252, 0.8);
        }

        .resources-grid {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(250px, 1fr));
          gap: var(--space-lg);
        }

        .resource-card {
          padding: var(--space-xl);
          background: rgba(255, 255, 255, 0.02);
          border: 1px solid rgba(255, 255, 255, 0.1);
          border-radius: 12px;
          text-decoration: none;
          color: inherit;
          transition: all 0.2s ease;
        }

        .resource-card:hover {
          border-color: var(--product-color);
          background: rgba(59, 130, 246, 0.1);
        }

        .resource-card svg {
          color: var(--product-color);
          margin-bottom: var(--space-md);
        }

        .resource-card h4 {
          margin-bottom: var(--space-xs);
          color: #f8fafc;
        }

        .resource-card p {
          color: rgba(248, 250, 252, 0.6);
          font-size: 0.85rem;
          margin: 0;
        }

        .cta-section {
          text-align: center;
          padding: var(--space-3xl);
          background: rgba(59, 130, 246, 0.05);
          border-radius: 16px;
          border: 1px solid rgba(59, 130, 246, 0.2);
        }

        .cta-section h2 {
          border: none;
          font-size: 2rem;
          margin-bottom: var(--space-xs);
        }

        .cta-subtitle {
          font-size: 1.25rem;
          color: var(--product-color-light);
          margin-bottom: var(--space-xl);
        }

        @media (max-width: 768px) {
          .stack-layer {
            flex-direction: column;
            align-items: flex-start;
            gap: var(--space-xs);
          }
        }
      `}</style>
    </div>
  );
}
