// @a0n/.ignored_gnosis/forkjoin-app/client/components/Presence/CursorOverlay.tsx
import { memo, useEffect, useState } from '@a0n/raect';
import type { PresenceCursor, Visitor } from '../../hooks/usePresence';

interface CursorOverlayProps {
  visitors: Visitor[];
}

interface ViewportState {
  width: number;
  height: number;
  scrollX: number;
  scrollY: number;
}

function readViewportState(): ViewportState {
  if (typeof window === 'undefined') {
    return {
      width: 1,
      height: 1,
      scrollX: 0,
      scrollY: 0,
    };
  }

  return {
    width: Math.max(window.innerWidth, 1),
    height: Math.max(window.innerHeight, 1),
    scrollX: window.scrollX,
    scrollY: window.scrollY,
  };
}

function resolveCursorPosition(
  cursor: PresenceCursor,
  viewport: ViewportState
): { x: number; y: number } {
  const hasViewportContext =
    typeof cursor.scrollX === 'number' ||
    typeof cursor.scrollY === 'number' ||
    typeof cursor.viewportWidth === 'number' ||
    typeof cursor.viewportHeight === 'number';

  if (!hasViewportContext) {
    return { x: cursor.x, y: cursor.y };
  }

  const remoteViewportWidth =
    typeof cursor.viewportWidth === 'number' && cursor.viewportWidth > 0
      ? cursor.viewportWidth
      : viewport.width;
  const remoteViewportHeight =
    typeof cursor.viewportHeight === 'number' && cursor.viewportHeight > 0
      ? cursor.viewportHeight
      : viewport.height;
  const remoteScrollX = typeof cursor.scrollX === 'number' ? cursor.scrollX : 0;
  const remoteScrollY = typeof cursor.scrollY === 'number' ? cursor.scrollY : 0;

  const absoluteX = remoteScrollX + cursor.x * remoteViewportWidth;
  const absoluteY = remoteScrollY + cursor.y * remoteViewportHeight;

  return {
    x: (absoluteX - viewport.scrollX) / viewport.width,
    y: (absoluteY - viewport.scrollY) / viewport.height,
  };
}

export const CursorOverlay = memo(function CursorOverlay({
  visitors,
}: CursorOverlayProps) {
  const [viewport, setViewport] = useState<ViewportState>(() =>
    readViewportState()
  );

  useEffect(() => {
    if (typeof window === 'undefined') {
      return;
    }

    let frameId: number | null = null;
    const scheduleSync = () => {
      if (frameId !== null) {
        return;
      }
      frameId = window.requestAnimationFrame(() => {
        frameId = null;
        setViewport(readViewportState());
      });
    };

    window.addEventListener('scroll', scheduleSync, { passive: true });
    window.addEventListener('resize', scheduleSync);

    return () => {
      if (frameId !== null) {
        window.cancelAnimationFrame(frameId);
      }
      window.removeEventListener('scroll', scheduleSync);
      window.removeEventListener('resize', scheduleSync);
    };
  }, []);

  // Filter to visitors with cursor positions
  const visibleCursors = visitors.filter((v) => v.cursor !== null);

  return (
    <div className="cursor-overlay" aria-hidden="true">
      {visibleCursors.map((visitor) => {
        if (!visitor.cursor) {
          return null;
        }
        const cursorPosition = resolveCursorPosition(visitor.cursor, viewport);
        return (
          <CursorIndicator
            key={visitor.id}
            x={cursorPosition.x}
            y={cursorPosition.y}
            color={visitor.color}
          />
        );
      })}

      <style>{`
        .cursor-overlay {
          position: fixed;
          inset: 0;
          pointer-events: none;
          z-index: 1000;
          overflow: hidden;
        }
      `}</style>
    </div>
  );
});

interface CursorIndicatorProps {
  x: number;
  y: number;
  color: string;
}

const CursorIndicator = memo(function CursorIndicator({
  x,
  y,
  color,
}: CursorIndicatorProps) {
  return (
    <div
      className="cursor-indicator"
      style={{
        left: `${x * 100}%`,
        top: `${y * 100}%`,
        backgroundColor: color,
        color: color,
      }}
    >
      <style>{`
        .cursor-indicator {
          position: absolute;
          width: 12px;
          height: 12px;
          border-radius: 50%;
          transform: translate(-50%, -50%);
          transition: left 0.1s ease-out, top 0.1s ease-out;
          pointer-events: none;
          opacity: 0.9;
          box-shadow: 0 0 10px currentColor, 0 0 20px currentColor;
        }

        .cursor-indicator::after {
          content: '';
          position: absolute;
          inset: -4px;
          border-radius: 50%;
          border: 2px solid currentColor;
          opacity: 0.4;
          animation: cursor-ring 2s ease-in-out infinite;
        }

        @keyframes cursor-ring {
          0%, 100% { transform: scale(1); opacity: 0.4; }
          50% { transform: scale(1.3); opacity: 0.2; }
        }
      `}</style>
    </div>
  );
});
