// @a0n/.ignored_gnosis/forkjoin-app/client/components/Presence/VisitorCount.tsx
import { memo } from '@a0n/raect';

interface VisitorCountProps {
  count: number;
  isConnected: boolean;
}

export const VisitorCount = memo(function VisitorCount({
  count,
  isConnected,
}: VisitorCountProps) {
  return (
    <div className="visitor-count">
      <span
        className={`status-dot ${isConnected ? 'connected' : 'disconnected'}`}
      />
      <span className="count">{count}</span>
      <span className="label">
        {count === 1 ? 'soul' : 'souls'} connected worldwide
      </span>

      <style>{`
        .visitor-count {
          display: inline-flex;
          align-items: center;
          gap: 0.5rem;
          padding: 0.5rem 1rem;
          background: rgba(10, 10, 15, 0.8);
          backdrop-filter: blur(10px);
          border: 1px solid rgba(99, 102, 241, 0.3);
          border-radius: 9999px;
          font-size: 0.875rem;
          margin-bottom: 1.5rem;
        }

        .status-dot {
          width: 8px;
          height: 8px;
          border-radius: 50%;
          background: var(--color-success);
        }

        .status-dot.connected {
          animation: pulse-dot 2s ease-in-out infinite;
        }

        .status-dot.disconnected {
          background: var(--color-warning);
        }

        @keyframes pulse-dot {
          0%, 100% { opacity: 0.6; transform: scale(1); }
          50% { opacity: 1; transform: scale(1.1); }
        }

        .count {
          font-weight: 600;
          color: var(--color-text);
          font-variant-numeric: tabular-nums;
        }

        .label {
          color: var(--color-text-secondary);
        }
      `}</style>
    </div>
  );
});
