// @a0n/.ignored_gnosis/gnosis-chaos/napi.ts
// Zero-copy NAPI binding for gnosis-chaos inference kernels.
//
// Node `Buffer` shares backing memory with `Float32Array` when constructed via
// `Buffer.from(view.buffer, view.byteOffset, view.byteLength)` -- no copy,
// no allocation. The Rust side reinterprets the bytes via `bytemuck::cast_slice`.
//
// Build: run `./build-napi.sh` from this directory; that produces
// `dist/gnosis_chaos.node`. If missing, the stub throws a descriptive error so
// callers can detect unbuilt environments and fall back to pure-JS kernels.

import { createRequire } from 'node:module';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

/**
 * Aggregated telemetry across all heads for a single call to
 * `forwardMedianShiftAttention`. Counts and masses are summed over heads;
 * `droppedMassFraction` is the arithmetic mean of per-head fractions.
 */
export interface MedianShiftTelemetry {
  readonly seqLen: number;
  readonly keptCount: number;
  readonly droppedCount: number;
  readonly zFull: number;
  readonly zDropped: number;
  readonly droppedMassFraction: number;
  readonly fellBackToFullSoftmax: boolean;
  readonly fallbackHeadCount: number;
}

type NativeModule = {
  forwardMultiheadAttention: (
    q: Buffer,
    k: Buffer,
    v: Buffer,
    out: Buffer,
    seqLen: number,
    headDim: number,
    numHeads: number,
    numKvHeads: number,
    scale: number,
  ) => void;
  forwardMedianShiftAttention: (
    q: Buffer,
    k: Buffer,
    v: Buffer,
    out: Buffer,
    seqLen: number,
    headDim: number,
    numHeads: number,
    numKvHeads: number,
    scale: number,
    droppedMassCeiling: number,
  ) => MedianShiftTelemetry;
  forwardFfnColumnTp: (
    x: Buffer,
    wUp: Buffer,
    wGate: Buffer,
    wDown: Buffer,
    out: Buffer,
    seqLen: number,
    hiddenDim: number,
    intermediateDim: number,
    numShards: number,
  ) => void;
  forwardMatvecF32: (
    matrix: Buffer,
    vector: Buffer,
    out: Buffer,
    rows: number,
    cols: number,
    softcap?: number,
  ) => void;
  renderTdaFrame: (
    width: number,
    height: number,
    time: number,
    betti1: number,
    activeNodesLow: number,
    activeNodesHigh: number,
    adjacencyMatrix: Buffer,
    profileArt: boolean,
    lod: number,
  ) => string;
};

function loadNative(): NativeModule | null {
  // Works under ESM, CJS, and ts-node. __dirname is absent in ESM.
  const here =
    typeof __dirname !== 'undefined'
      ? __dirname
      : path.dirname(fileURLToPath(import.meta.url));
  const binaryPath = path.join(here, 'dist', 'gnosis_chaos.node');
  try {
    const req =
      typeof require !== 'undefined' ? require : createRequire(import.meta.url);
    return req(binaryPath) as NativeModule;
  } catch {
    return null;
  }
}

const native = loadNative();

function asBuffer(view: Float32Array): Buffer {
  return Buffer.from(view.buffer, view.byteOffset, view.byteLength);
}

function assertNative(native: NativeModule | null): asserts native is NativeModule {
  if (!native) {
    throw new Error(
      'gnosis-chaos native binding not built; run build-napi.sh',
    );
  }
}

/**
 * Multi-head attention forward pass over a single decode-time query per head.
 *
 * Zero-copy: the Float32Array views are passed as Node Buffers that share the
 * same backing ArrayBuffer. Rust parallelises heads via rayon.
 *
 * Shapes (row-major f32):
 *   q:   [numHeads   * headDim]
 *   k:   [numKvHeads * seqLen * headDim]
 *   v:   [numKvHeads * seqLen * headDim]
 *   out: [numHeads   * headDim]  -- overwritten
 */
export function forwardMultiheadAttention(
  q: Float32Array,
  k: Float32Array,
  v: Float32Array,
  out: Float32Array,
  seqLen: number,
  headDim: number,
  numHeads: number,
  numKvHeads: number,
  scale: number,
): void {
  assertNative(native);
  native.forwardMultiheadAttention(
    asBuffer(q),
    asBuffer(k),
    asBuffer(v),
    asBuffer(out),
    seqLen,
    headDim,
    numHeads,
    numKvHeads,
    scale,
  );
}

/**
 * Median-shift attention (Angle 4). Drops positions at/below the row median
 * before softmax, renormalises over the kept set, and falls back to full
 * softmax per-head when the dropped mass fraction exceeds `droppedMassCeiling`.
 *
 * Returns aggregate telemetry (counts summed over heads; fractions averaged).
 *
 * Shapes (row-major f32):
 *   q:   [numHeads   * headDim]
 *   k:   [numKvHeads * seqLen * headDim]
 *   v:   [numKvHeads * seqLen * headDim]
 *   out: [numHeads   * headDim]  -- overwritten
 */
export function forwardMedianShiftAttention(
  q: Float32Array,
  k: Float32Array,
  v: Float32Array,
  out: Float32Array,
  seqLen: number,
  headDim: number,
  numHeads: number,
  numKvHeads: number,
  scale: number,
  droppedMassCeiling: number = 0.15,
): MedianShiftTelemetry {
  assertNative(native);
  return native.forwardMedianShiftAttention(
    asBuffer(q),
    asBuffer(k),
    asBuffer(v),
    asBuffer(out),
    seqLen,
    headDim,
    numHeads,
    numKvHeads,
    scale,
    droppedMassCeiling,
  );
}

/**
 * Column-parallel SwiGLU feed-forward pass (Angle 5).
 *
 * The `intermediateDim` axis is sharded across `numShards` rayon workers.
 * Each shard computes a stripe of (up, gate) then its partial contribution
 * to `out`; the mesh fold is a sequential rayon reduce across shards.
 *
 * Shapes (row-major f32):
 *   x:     [seqLen,          hiddenDim]
 *   wUp:   [hiddenDim,       intermediateDim]
 *   wGate: [hiddenDim,       intermediateDim]
 *   wDown: [intermediateDim, hiddenDim]
 *   out:   [seqLen,          hiddenDim]  -- overwritten
 */
export function forwardFfnColumnTp(
  x: Float32Array,
  wUp: Float32Array,
  wGate: Float32Array,
  wDown: Float32Array,
  out: Float32Array,
  seqLen: number,
  hiddenDim: number,
  intermediateDim: number,
  numShards: number,
): void {
  assertNative(native);
  native.forwardFfnColumnTp(
    asBuffer(x),
    asBuffer(wUp),
    asBuffer(wGate),
    asBuffer(wDown),
    asBuffer(out),
    seqLen,
    hiddenDim,
    intermediateDim,
    numShards,
  );
}

/**
 * Dense matrix-vector multiply: `out = matrix @ vector` where matrix is
 * `rows × cols` row-major Float32 and vector is length `cols`. `out` is
 * overwritten in place with length `rows`. Parallelism over rows via
 * rayon; each row is an independent dot product of length `cols`.
 *
 * Hot-path replacement for `SIMDFallback.matVec`. A single Gemma-4-31B
 * transformer layer calls matVec 7× (Q/K/V/O + FFN up/gate/down); the
 * full model is 60 layers × 7 = 420 matVec calls per token.
 */
export function forwardMatvecF32(
  matrix: Float32Array,
  vector: Float32Array,
  out: Float32Array,
  rows: number,
  cols: number,
  softcap?: number,
): void {
  assertNative(native);
  native.forwardMatvecF32(
    asBuffer(matrix),
    asBuffer(vector),
    asBuffer(out),
    rows,
    cols,
    softcap,
  );
}

/**
 * Handles the gnosis render Tda Frame workflow.
 */
export function renderTdaFrame(
  width: number,
  height: number,
  time: number,
  betti1: number,
  activeNodesLow: number,
  activeNodesHigh: number,
  adjacencyMatrix: BigUint64Array,
  profileArt: boolean,
  lod: number = 0,
): string {
  if (NATIVE_BINDING_AVAILABLE && native) {
    // Adjacency matrix is 64 * u64 (8 bytes each) = 512 bytes
    const buffer = Buffer.from(
      adjacencyMatrix.buffer,
      adjacencyMatrix.byteOffset,
      adjacencyMatrix.byteLength,
    );
    return native.renderTdaFrame(
      width,
      height,
      time,
      betti1,
      activeNodesLow,
      activeNodesHigh,
      buffer,
      profileArt,
      lod,
    );
  }

  return '';
}

export const NATIVE_BINDING_AVAILABLE: boolean = native !== null;
