// @a0n/.ignored_gnosis/.workspace-stubs/twokeys/index.js
function numericData(data) {
  return Array.isArray(data) ? data.filter((value) => Number.isFinite(value)) : [];
}

function mean(values) {
  return values.length === 0
    ? 0
    : values.reduce((sum, value) => sum + value, 0) / values.length;
}

export class Series {
  constructor(options = {}) {
    this.data = numericData(options.data);
  }

  describe() {
    const sorted = [...this.data].sort((left, right) => left - right);
    const average = mean(sorted);
    return {
      count: sorted.length,
      min: sorted[0] ?? 0,
      max: sorted.at(-1) ?? 0,
      mean: average,
      average,
      sum: sorted.reduce((total, value) => total + value, 0),
    };
  }
}

export class Points {
  constructor(options = {}) {
    this.data = Array.isArray(options.data) ? options.data : [];
  }

  describe() {
    return {
      count: this.data.length,
      dimensions: this.data[0]?.length ?? 0,
      centroid: this.data[0]?.map((_, index) =>
        mean(this.data.map((point) => point[index]).filter((value) => Number.isFinite(value)))
      ) ?? [],
    };
  }
}

/**
 * Handles the gnosis graph Eda workflow.
 */
export function graphEda(nodes = [], edges = []) {
  return {
    nodeCount: nodes.length,
    edgeCount: edges.length,
    density:
      nodes.length <= 1 ? 0 : edges.length / Math.max(1, nodes.length * (nodes.length - 1)),
  };
}

/**
 * Handles the gnosis graph Outliers workflow.
 */
export function graphOutliers() {
  return [];
}

export default {
  Points,
  Series,
  graphEda,
  graphOutliers,
};
