// @a0n/.ignored_gnosis/src/betty/buleyean.ts
/**
 * Buleyean probability computation with positivity guarantee.
 *
 * Takes a VoidBoundary and returns a complement distribution where
 * all weights > 0. The irreducible +1 sliver ensures no fold ever
 * produces probability zero.
 *
 * Exported for both compile-time and runtime use.
 */

import type { VoidBoundary } from '../void';
import { buleyeanWeight } from '../void';

export interface BuleyeanDistribution {
  /** Per-dimension weights (all > 0) */
  weights: number[];
  /** Normalized probabilities (all > 0, sum to 1) */
  probabilities: number[];
  /** Whether positivity holds (always true by construction) */
  positive: boolean;
  /** Minimum weight (always >= 1 due to sliver) */
  minimumWeight: number;
}

/**
 * Compute the Buleyean distribution from a void boundary.
 * The +1 sliver in buleyeanWeight guarantees all weights >= 1.
 */
export function computeBuleyeanDistribution(
  boundary: VoidBoundary,
  rounds: number = 0
): BuleyeanDistribution {
  const T = rounds > 0 ? rounds : boundary.totalEntries;
  const weights = boundary.counts.map((v) => buleyeanWeight(T, v));
  const sum = weights.reduce((a, b) => a + b, 0);
  const probabilities =
    sum > 0
      ? weights.map((w) => w / sum)
      : weights.map(() => 1 / weights.length);
  const minimumWeight = weights.length > 0 ? Math.min(...weights) : 1;

  return {
    weights,
    probabilities,
    positive: minimumWeight >= 1,
    minimumWeight,
  };
}

/**
 * Check if a fold would produce zero probability on any branch.
 * Returns the indices of branches that would be zeroed.
 */
export function checkFoldZeroProbability(
  boundary: VoidBoundary,
  rounds: number = 0
): number[] {
  const dist = computeBuleyeanDistribution(boundary, rounds);
  // By construction, no branch can be zero (sliver guarantee).
  // This function exists for compile-time verification.
  return dist.weights.map((w, i) => (w <= 0 ? i : -1)).filter((i) => i >= 0);
}
