// @a0n/.ignored_gnosis/src/betty/buleyean-feedback.ts
/**
 * Buleyean Rejection Collector for Betty Compiler -- code diffusion self-healing.
 *
 * Hooks into the BettyCompiler's diagnostic output and the semantic healer
 * to accumulate rejection records in Buleyean void-boundary form. Each
 * DiagnosticCode is a dimension; severity determines rejection magnitude.
 * The complement distribution over diagnostic codes reveals which error
 * types the system has learned to avoid -- the Buleyean complement of
 * compiler failure.
 *
 * Unhealable edges from the semantic healer are captured as stronger
 * rejections (magnitude 5), reinforcing the void where healing fails.
 *
 * The Hope Certificate's elimination rate is tracked over time to show
 * convergence of the self-healing loop.
 */

import type { VoidBoundary } from '../void';
import {
  createVoidBoundary,
  updateVoidBoundary,
  complementDistribution,
  buleyeanWeight,
  shannonEntropy,
} from '../void';
import type {
  Diagnostic,
  DiagnosticCode,
  BettyParseResult,
} from './compiler';
import type { HealingResult, UnhealableEdge } from './semantic-healer';
import type { HopeCertificate } from './semantic-hope';

// ─── Diagnostic Code Registry ───────────────────────────────────────────────

/**
 * All 38 DiagnosticCode values, each mapped to a dimension index.
 * One dimension per code in the CompilerVoidBoundary.
 */
export const DIAGNOSTIC_CODES: readonly DiagnosticCode[] = [
  'ERR_BETA_UNBOUNDED',
  'ERR_SPECTRAL_EXPLOSION',
  'ERR_DRIFT_POSITIVE',
  'ERR_REPAIR_DEBT_LEAK',
  'ERR_CONTINUOUS_WITNESS_INVALID',
  'VOID_ETA_UNBOUNDED',
  'VOID_EXPLORATION_UNBOUNDED',
  'VOID_COMPLEMENT_DISTRIBUTION_INVALID',
  'VOID_BOUNDARY_NON_MONOTONE',
  'VOID_GAIT_TRANSITION_INVALID',
  'VOID_METACOG_MISSING_CONVERGENCE',
  'VOID_CONSERVATION_VIOLATED',
  'VOID_TRACED_MONOIDAL_VIOLATED',
  'VOID_METACOG_PARAMETER_DRIFT',
  'ETHICS_FOLD_WITHOUT_EVIDENCE',
  'ETHICS_FOLD_DESTROYS_ALL',
  'ETHICS_VENT_WITHOUT_BOUNDARY',
  'ETHICS_RACE_PREMATURE_COLLAPSE',
  'ETHICS_FORK_ZERO_OPTIONS',
  'ETHICS_MISSING_VENT_PATH',
  'ETHICS_IRREVERSIBLE_ON_OTHERS_VOID',
  'ETHICS_NO_TRACE',
  'ERR_COARSENING_UNCOVERED_NODE',
  'ERR_COARSENING_SERVICE_NOT_POSITIVE',
  'ERR_COARSENING_DRIFT_POSITIVE',
  'ERR_COARSENING_SYMBOLIC_RATE',
  'ERR_DEFICIT_NONZERO',
  'THERMO_FIRST_LAW_VIOLATED',
  'BULE_POSITIVITY_VIOLATED',
  'BULE_EXHAUSTIVENESS_INCOMPLETE',
  'INFO_DEFICIT_SUBGRAPH',
  'ENTANGLE_DIM_MISMATCH',
  'ALEPH_INCOMPLETE',
  'ENTROPY_REVERSAL_DETECTED',
  'SEMANTIC_TYPE_INCOMPATIBLE',
  'SEMANTIC_PROOF_OBLIGATION',
  'SEMANTIC_FOLD_MISMATCH',
  'LAW_IMPOSSIBILITY_OF_ZERO',
  'LAW_STRICT_ORDERING',
  'LAW_UNIVERSAL_SANDWICH',
  'LAW_CAVE_DEFICIT',
  'LAW_CONSERVATION',
  'LAW_SORITES_SHARPNESS',
  'LAW_CHAIN_TERMINATION',
  'GOD_FORMULA_EXISTENCE',
  'GOD_FORMULA_PERSISTENCE',
  'GOD_FORMULA_LEARNING',
  'GOD_FORMULA_DISCRIMINATION',
  'GOD_FORMULA_BOUNDEDNESS',
  'GOD_FORMULA_CONVERGENCE',
  'GOD_FORMULA_UNCERTAINTY',
  'GOD_FORMULA_COHERENCE',
  'GOD_FORMULA_HOPE',
  'CTRL_INDIVIDUAL',
  'CTRL_AGGREGATE',
  'CTRL_SPREAD',
  'CTRL_MEAN',
] as const;

/** Map from DiagnosticCode to dimension index for O(1) lookup. */
const CODE_TO_DIM: ReadonlyMap<DiagnosticCode, number> = new Map(
  DIAGNOSTIC_CODES.map((code, idx) => [code, idx])
);

/** Number of diagnostic code dimensions. */
export const DIAGNOSTIC_DIMENSIONS = DIAGNOSTIC_CODES.length;

// ─── Severity Weights ───────────────────────────────────────────────────────

/** Severity-weighted rejection magnitudes. */
const SEVERITY_WEIGHT: Record<Diagnostic['severity'], number> = {
  error: 3,
  warning: 1,
  info: 0,
};

/** Unhealable edges are stronger rejections than individual diagnostics. */
const UNHEALABLE_MAGNITUDE = 5;

// ─── Rejection Record ───────────────────────────────────────────────────────

/**
 * A single Buleyean RL training record capturing one compilation attempt.
 * JSONL-compatible: each record serializes to one line.
 */
export interface RejectionRecord {
  /** The .gg topology source that was compiled. */
  prompt: string;
  /** Array of diagnostic messages from this compilation attempt. */
  rejected_responses: string[];
  /** Severity-weighted rejection count for this attempt. */
  rejection_counts: number;
  /** Which compilation round this represents (1-indexed). */
  total_rounds: number;
  /** Per-diagnostic-code rejection counts for this attempt. */
  per_code_counts: Record<string, number>;
  /** Unhealable edge descriptions, if any. */
  unhealable_edges: string[];
  /** Hope Certificate elimination rate, if available. */
  elimination_rate: number | null;
  /** Timestamp of this record. */
  timestamp: string;
}

// ─── Elimination Rate Tracker ───────────────────────────────────────────────

interface EliminationSnapshot {
  round: number;
  rate: number;
  healed: number;
  unhealable: number;
}

// ─── CompilerVoidBoundary ───────────────────────────────────────────────────

/**
 * A VoidBoundary specialized for the compiler's diagnostic space.
 * Accumulates rejection counts per DiagnosticCode across compilation
 * attempts. The complement distribution reveals which error types
 * the system has learned to avoid.
 */
export interface CompilerVoidBoundary {
  /** The underlying void boundary with one dimension per DiagnosticCode. */
  boundary: VoidBoundary;
  /** Total compilation rounds observed. */
  totalRounds: number;
  /** All rejection records accumulated. */
  records: RejectionRecord[];
  /** Elimination rate history from Hope Certificates. */
  eliminationHistory: EliminationSnapshot[];
}

/**
 * Create a fresh CompilerVoidBoundary.
 */
export function createCompilerVoidBoundary(): CompilerVoidBoundary {
  return {
    boundary: createVoidBoundary(DIAGNOSTIC_DIMENSIONS),
    totalRounds: 0,
    records: [],
    eliminationHistory: [],
  };
}

// ─── Core Feedback Loop ─────────────────────────────────────────────────────

/**
 * Record diagnostics from a compilation attempt into the void boundary.
 * Each diagnostic with a known code updates the corresponding dimension
 * with severity-weighted magnitude.
 *
 * Call this after each BettyCompiler.parse() invocation.
 */
export function recordCompilationDiagnostics(
  cvb: CompilerVoidBoundary,
  source: string,
  diagnostics: Diagnostic[]
): void {
  cvb.totalRounds++;

  const perCodeCounts: Record<string, number> = {};
  const rejectedResponses: string[] = [];
  let totalRejection = 0;

  for (const diag of diagnostics) {
    const magnitude = SEVERITY_WEIGHT[diag.severity];
    if (magnitude === 0) continue;

    rejectedResponses.push(
      `[${diag.severity}] ${diag.code ?? 'UNTYPED'}: ${diag.message}`
    );

    if (diag.code) {
      const dim = CODE_TO_DIM.get(diag.code);
      if (dim !== undefined) {
        updateVoidBoundary(cvb.boundary, dim, magnitude);
        perCodeCounts[diag.code] = (perCodeCounts[diag.code] ?? 0) + magnitude;
        totalRejection += magnitude;
      }
    }
  }

  cvb.records.push({
    prompt: source,
    rejected_responses: rejectedResponses,
    rejection_counts: totalRejection,
    total_rounds: cvb.totalRounds,
    per_code_counts: perCodeCounts,
    unhealable_edges: [],
    elimination_rate: null,
    timestamp: new Date().toISOString(),
  });
}

/**
 * Record unhealable edges from the semantic healer as stronger rejections.
 * Unhealable edges indicate structural failures that no converter can fix --
 * these deserve higher rejection magnitude to steer future topologies away
 * from the patterns that produce them.
 *
 * Call this after healTopology() when unhealable edges remain.
 */
export function recordUnhealableEdges(
  cvb: CompilerVoidBoundary,
  unhealableEdges: UnhealableEdge[]
): void {
  if (unhealableEdges.length === 0) return;

  // Unhealable edges map to SEMANTIC_TYPE_INCOMPATIBLE dimension
  const dim = CODE_TO_DIM.get('SEMANTIC_TYPE_INCOMPATIBLE');
  if (dim === undefined) return;

  for (const edge of unhealableEdges) {
    updateVoidBoundary(cvb.boundary, dim, UNHEALABLE_MAGNITUDE);
  }

  // Append unhealable info to the most recent record
  const lastRecord = cvb.records[cvb.records.length - 1];
  if (lastRecord) {
    lastRecord.unhealable_edges = unhealableEdges.map(
      (e) => `${e.sourceNode} -> ${e.targetNode}: ${e.reason}`
    );
    lastRecord.rejection_counts +=
      unhealableEdges.length * UNHEALABLE_MAGNITUDE;
    lastRecord.per_code_counts['SEMANTIC_TYPE_INCOMPATIBLE'] =
      (lastRecord.per_code_counts['SEMANTIC_TYPE_INCOMPATIBLE'] ?? 0) +
      unhealableEdges.length * UNHEALABLE_MAGNITUDE;
  }
}

/**
 * Record the Hope Certificate's elimination rate for convergence tracking.
 * The elimination rate measures healed / (healed + unhealable) -- as this
 * approaches 1.0 over rounds, the system is learning to produce healable
 * topologies.
 *
 * Call this after generateHopeCertificate().
 */
export function recordEliminationRate(
  cvb: CompilerVoidBoundary,
  hopeCertificate: HopeCertificate
): void {
  const g4 = hopeCertificate.guarantees.confusionIsEliminable;
  const snapshot: EliminationSnapshot = {
    round: cvb.totalRounds,
    rate: g4.eliminationRate,
    healed: g4.healed,
    unhealable: g4.unhealable,
  };
  cvb.eliminationHistory.push(snapshot);

  // Update the most recent record with the elimination rate
  const lastRecord = cvb.records[cvb.records.length - 1];
  if (lastRecord) {
    lastRecord.elimination_rate = g4.eliminationRate;
  }
}

// ─── Convenience: Full Parse Result Hook ────────────────────────────────────

/**
 * Process an entire BettyParseResult in one call. Captures diagnostics,
 * unhealable edges (if healing was performed), and the Hope Certificate
 * elimination rate.
 *
 * This is the primary hook point for the compiler loop:
 *   const result = compiler.parse(source);
 *   recordParseResult(cvb, source, result, healingResult);
 */
export function recordParseResult(
  cvb: CompilerVoidBoundary,
  source: string,
  result: BettyParseResult,
  healingResult?: HealingResult
): void {
  // Step 1: Record all diagnostics
  recordCompilationDiagnostics(cvb, source, result.diagnostics);

  // Step 2: Record unhealable edges if healing was attempted
  if (healingResult) {
    recordUnhealableEdges(cvb, healingResult.unhealable);
  }

  // Step 3: Record Hope Certificate elimination rate
  if (result.hopeCertificate) {
    recordEliminationRate(cvb, result.hopeCertificate);
  }
}

// ─── Complement Distribution ────────────────────────────────────────────────

/**
 * The Buleyean complement distribution over diagnostic codes.
 * Where the void has accumulated *least*, the complement peaks --
 * these are the error types the system has learned to avoid.
 *
 * Returns a labeled distribution mapping each DiagnosticCode to its
 * complement probability.
 */
export function getComplementDistribution(
  cvb: CompilerVoidBoundary,
  eta: number = 3.0
): Map<DiagnosticCode, number> {
  const dist = complementDistribution(cvb.boundary, eta);
  const result = new Map<DiagnosticCode, number>();
  for (let i = 0; i < DIAGNOSTIC_CODES.length; i++) {
    result.set(DIAGNOSTIC_CODES[i], dist[i]);
  }
  return result;
}

/**
 * Get the raw complement distribution as a plain array (same order as
 * DIAGNOSTIC_CODES). Useful for downstream consumers that want unlabeled
 * probability vectors.
 */
export function getComplementDistributionArray(
  cvb: CompilerVoidBoundary,
  eta: number = 3.0
): number[] {
  return complementDistribution(cvb.boundary, eta);
}

// ─── Buleyean Weights ───────────────────────────────────────────────────────

/**
 * Per-dimension Buleyean weights. Weight = T - min(v_i, T) + 1.
 * Higher weight means less rejection in that dimension -- the system
 * is more confident about avoiding that error type.
 */
export function getDiagnosticWeights(
  cvb: CompilerVoidBoundary
): Map<DiagnosticCode, number> {
  const T = cvb.totalRounds > 0 ? cvb.totalRounds : cvb.boundary.totalEntries;
  const result = new Map<DiagnosticCode, number>();
  for (let i = 0; i < DIAGNOSTIC_CODES.length; i++) {
    result.set(DIAGNOSTIC_CODES[i], buleyeanWeight(T, cvb.boundary.counts[i]));
  }
  return result;
}

// ─── Elimination Rate Analysis ──────────────────────────────────────────────

/**
 * Get the elimination rate trend over time.
 * Returns the full history and a simple linear trend indicator.
 */
export function getEliminationTrend(cvb: CompilerVoidBoundary): {
  history: EliminationSnapshot[];
  improving: boolean;
  currentRate: number;
  averageRate: number;
} {
  const history = cvb.eliminationHistory;
  if (history.length === 0) {
    return {
      history: [],
      improving: false,
      currentRate: 1.0,
      averageRate: 1.0,
    };
  }

  const currentRate = history[history.length - 1].rate;
  const averageRate =
    history.reduce((sum, s) => sum + s.rate, 0) / history.length;

  // Compare first half average to second half average
  const mid = Math.floor(history.length / 2);
  const firstHalf = history.slice(0, mid || 1);
  const secondHalf = history.slice(mid);
  const firstAvg = firstHalf.reduce((s, h) => s + h.rate, 0) / firstHalf.length;
  const secondAvg =
    secondHalf.reduce((s, h) => s + h.rate, 0) / secondHalf.length;

  return {
    history,
    improving: secondAvg >= firstAvg,
    currentRate,
    averageRate,
  };
}

// ─── JSONL Serialization ────────────────────────────────────────────────────

/**
 * Serialize accumulated rejection data as Buleyean RL training records
 * in JSONL-compatible format. Each line is one JSON object representing
 * a single compilation attempt with its rejection signal.
 */
export function toRejectionRecords(cvb: CompilerVoidBoundary): string {
  return cvb.records.map((record) => JSON.stringify(record)).join('\n');
}

/**
 * Serialize a summary record that captures the aggregate void boundary
 * state, complement distribution, and elimination trend. This is the
 * "learned model" -- the Buleyean complement of all compiler failures
 * observed so far.
 */
export function toSummaryRecord(
  cvb: CompilerVoidBoundary,
  eta: number = 3.0
): string {
  const dist = getComplementDistribution(cvb, eta);
  const trend = getEliminationTrend(cvb);
  const entropy = shannonEntropy(complementDistribution(cvb.boundary, eta));

  const summary = {
    type: 'buleyean_compiler_summary',
    total_rounds: cvb.totalRounds,
    total_rejections: cvb.boundary.totalEntries,
    dimensions: DIAGNOSTIC_DIMENSIONS,
    per_code_totals: Object.fromEntries(
      DIAGNOSTIC_CODES.map((code, i) => [code, cvb.boundary.counts[i]])
    ),
    complement_distribution: Object.fromEntries(dist),
    complement_entropy: entropy,
    elimination_trend: {
      improving: trend.improving,
      current_rate: trend.currentRate,
      average_rate: trend.averageRate,
      snapshots: trend.history.length,
    },
    timestamp: new Date().toISOString(),
  };

  return JSON.stringify(summary);
}
