// @a0n/.ignored_gnosis/src/betty/compiler.ts
import { QuantumWasmBridge } from './quantum/bridge';
import { injectSensitiveZkEnvelopes } from '../auth/auto-zk';
import { lowerUfcsSource } from '../ufcs';
import type { StabilityReport } from './stability';
import type { OptimizerResult } from './optimizer';
import type { CoarseningSynthesisResult } from './coarsen';
import type { SemanticCompatibilityResult } from './semantic-compatibility';
import type { HopeCertificate } from './semantic-hope';
import {
  lowerLegaleseToGg,
  type LegaleseLoweringOptions,
  type LegaleseLoweringResult,
} from './legalese';
import {
  parseNodeDeclarations,
  parseEdgeDeclarations,
  parseProperties as parsePropertiesShared,
  parseNodeRefsFromEdgeGroup,
  IMPERATIVE_KEYWORD_REGEX,
} from './parse-utils';
import { runAllVerificationPasses } from './verify';

export interface ASTNode {
  id: string;
  labels: string[];
  properties: Record<string, string>;
}

export interface ASTEdge {
  sourceIds: string[];
  targetIds: string[];
  type: string; // FORK, RACE, FOLD, VENT, PROCESS, COLLAPSE, TUNNEL, SLIVER, OBSERVE, LAMINAR, METACOG
  properties: Record<string, string>;
}

export interface Diagnostic {
  line: number;
  column: number;
  code?: DiagnosticCode;
  message: string;
  severity: 'error' | 'warning' | 'info';
}

export type DiagnosticCode =
  | 'ERR_BETA_UNBOUNDED'
  | 'ERR_SPECTRAL_EXPLOSION'
  | 'ERR_DRIFT_POSITIVE'
  | 'ERR_REPAIR_DEBT_LEAK'
  | 'ERR_CONTINUOUS_WITNESS_INVALID'
  | 'VOID_ETA_UNBOUNDED'
  | 'VOID_EXPLORATION_UNBOUNDED'
  | 'VOID_COMPLEMENT_DISTRIBUTION_INVALID'
  | 'VOID_BOUNDARY_NON_MONOTONE'
  | 'VOID_GAIT_TRANSITION_INVALID'
  | 'VOID_METACOG_MISSING_CONVERGENCE'
  | 'VOID_CONSERVATION_VIOLATED'
  | 'VOID_TRACED_MONOIDAL_VIOLATED'
  | 'VOID_METACOG_PARAMETER_DRIFT'
  | 'ETHICS_FOLD_WITHOUT_EVIDENCE'
  | 'ETHICS_FOLD_DESTROYS_ALL'
  | 'ETHICS_VENT_WITHOUT_BOUNDARY'
  | 'ETHICS_RACE_PREMATURE_COLLAPSE'
  | 'ETHICS_FORK_ZERO_OPTIONS'
  | 'ETHICS_MISSING_VENT_PATH'
  | 'ETHICS_IRREVERSIBLE_ON_OTHERS_VOID'
  | 'ETHICS_NO_TRACE'
  | 'ERR_COARSENING_UNCOVERED_NODE'
  | 'ERR_COARSENING_SERVICE_NOT_POSITIVE'
  | 'ERR_COARSENING_DRIFT_POSITIVE'
  | 'ERR_COARSENING_SYMBOLIC_RATE'
  | 'ERR_DEFICIT_NONZERO'
  | 'THERMO_FIRST_LAW_VIOLATED'
  | 'BULE_POSITIVITY_VIOLATED'
  | 'BULE_EXHAUSTIVENESS_INCOMPLETE'
  | 'INFO_DEFICIT_SUBGRAPH'
  | 'ENTANGLE_DIM_MISMATCH'
  | 'ALEPH_INCOMPLETE'
  | 'ENTROPY_REVERSAL_DETECTED'
  | 'SEMANTIC_TYPE_INCOMPATIBLE'
  | 'SEMANTIC_PROOF_OBLIGATION'
  | 'SEMANTIC_FOLD_MISMATCH'
  | 'LAW_IMPOSSIBILITY_OF_ZERO'
  | 'LAW_STRICT_ORDERING'
  | 'LAW_UNIVERSAL_SANDWICH'
  | 'LAW_CAVE_DEFICIT'
  | 'LAW_CONSERVATION'
  | 'LAW_SORITES_SHARPNESS'
  | 'LAW_CHAIN_TERMINATION'
  | 'GOD_FORMULA_EXISTENCE'
  | 'GOD_FORMULA_PERSISTENCE'
  | 'GOD_FORMULA_LEARNING'
  | 'GOD_FORMULA_DISCRIMINATION'
  | 'GOD_FORMULA_BOUNDEDNESS'
  | 'GOD_FORMULA_CONVERGENCE'
  | 'GOD_FORMULA_UNCERTAINTY'
  | 'GOD_FORMULA_COHERENCE'
  | 'GOD_FORMULA_HOPE'
  | 'CTRL_INDIVIDUAL'
  | 'CTRL_AGGREGATE'
  | 'CTRL_SPREAD'
  | 'CTRL_MEAN'
  | 'LAYOUT_TAILWIND_CONFLICT'
  | 'LAYOUT_FLEX_OVERFLOW'
  | 'LAYOUT_POSITION_ESCAPE'
  | 'LAYOUT_RESPONSIVE_NON_MONOTONE'
  | 'LAYOUT_MARGIN_COLLAPSE'
  | 'LAYOUT_MISSING_FLEX_PARENT'
  | 'LAYOUT_OVERFLOW_UNHANDLED'
  | 'SEO_HEADING_HIERARCHY'
  | 'SEO_MISSING_META'
  | 'SEO_MISSING_STRUCTURED_DATA'
  | 'A11Y_MISSING_ALT'
  | 'A11Y_MISSING_LABEL'
  | 'A11Y_COLOR_CONTRAST'
  | 'A11Y_MISSING_LANG'
  | 'BOUNDARY_SERVER_IMPORT_IN_CLIENT'
  | 'BOUNDARY_CLIENT_IMPORT_IN_SERVER'
  | 'BOUNDARY_SECRET_IN_CLIENT'
  | 'CACHE_IMMUTABLE_MUTABLE_RESOURCE'
  | 'CACHE_NO_CACHE_HEADERS'
  | 'SLOP_EMOJI_IN_CODE'
  | 'SLOP_CONSOLE_LOG_IN_PROD'
  | 'SLOP_TODO_IN_PROD'
  | 'SLOP_HARDCODED_URL';

export interface GraphAST {
  nodes: Map<string, ASTNode>;
  edges: ASTEdge[];
}

export interface BettyParseResult {
  ast: GraphAST | null;
  output: string;
  b1: number;
  diagnostics: Diagnostic[];
  buleyMeasure: number;
  stability: StabilityReport | null;
  optimizer: OptimizerResult | null;
  coarseningSynthesis: CoarseningSynthesisResult | null;
  voidDimensions: number;
  landauerHeat: number;
  deficit: DeficitAnalysis | null;
  semanticCompatibility: SemanticCompatibilityResult | null;
  hopeCertificate: HopeCertificate | null;
  /** Clockwork self-verification verdict */
  clockwork: import('../universe/index.js').ClockworkVerdict;
  /** Universe certificate emitted by the God Formula kernel. */
  universe: import('../universe/index.js').UniverseCertificate;
  /** Packed knot binary wire format (2-bit crossing stream). */
  knotBinary: Uint8Array | null;
}

export interface BettyLegaleseParseResult extends BettyParseResult {
  ggSource: string;
  lowering: LegaleseLoweringResult;
}

export interface DeficitAnalysis {
  perNode: Map<string, number>;
  totalDeficit: number;
  diagnostics: Diagnostic[];
}

const NATIVE_TAGGED_NODE_CASES: Record<string, readonly string[]> = {
  Result: ['ok', 'err'],
  Option: ['some', 'none'],
};

const STRUCTURED_CONCURRENCY_FAILURES = new Set(['cancel', 'vent', 'shield']);
const GAIT_ORDER = ['stand', 'trot', 'canter', 'gallop'];
const TRACE_SUPPORTING_EDGE_TYPES = new Set([
  'MEASURE',
  'METACOG',
  'RACE',
  'SLIVER',
]);
const IMPLICIT_VENT_EDGE_TYPES = new Set(['RACE', 'SLIVER']);

export class BettyCompiler {
  private b1 = 0;
  private ast: GraphAST = { nodes: new Map(), edges: [] };
  private logs: string[] = [];
  private diagnostics: Diagnostic[] = [];
  private stability: StabilityReport | null = null;
  private optimizerResult: OptimizerResult | null = null;
  private coarseningSynthesisResult: CoarseningSynthesisResult | null = null;
  private semanticResult: SemanticCompatibilityResult | null = null;
  private wasmBridge: QuantumWasmBridge;

  constructor() {
    this.wasmBridge = new QuantumWasmBridge();
  }

  public getBettiNumber(): number {
    return this.b1;
  }

  public getAST(): GraphAST {
    return this.ast;
  }

  public getLogs(): string[] {
    return this.logs;
  }

  public getDiagnostics(): Diagnostic[] {
    return this.diagnostics;
  }

  public getStability(): StabilityReport | null {
    return this.stability;
  }

  /**
   * Buley Measurement: Topological Complexity Score
   * Calculates the entropy of the covering space.
   */
  public getBuleyMeasurement(): number {
    const pathComplexity = this.ast.edges.reduce((acc, edge) => {
      return acc + edge.sourceIds.length * edge.targetIds.length;
    }, 0);
    return this.b1 * 1.5 + pathComplexity * 0.5;
  }

  /**
   * Check if a node has semantic predicate annotations.
   * Supported predicates: invertible, idempotent, total-function, monotone, valid-json.
   *
   * Usage in .gg source:
   *   (transform: Statement { semantic_predicates='["invertible","total-function"]' })
   *   (compress: Statement { semantic_predicates='[{"kind":"valid-json","schema":{"type":"object"}}]' })
   */
  private validateSemanticPredicates(): void {
    for (const [nodeId, node] of this.ast.nodes) {
      const predStr = node.properties['semantic_predicates'];
      if (!predStr) continue;

      try {
        const preds = JSON.parse(predStr);
        if (!Array.isArray(preds)) {
          this.diagnostics.push({
            line: 0,
            column: 0,
            message: `Node '${nodeId}': semantic_predicates must be a JSON array`,
            severity: 'warning',
          });
          continue;
        }

        // Validate each predicate.
        for (const pred of preds) {
          if (typeof pred === 'string') {
            // Shorthand: just the predicate name.
            const validNames = [
              'invertible',
              'idempotent',
              'total-function',
              'monotone',
              'valid-json',
            ];
            if (!validNames.includes(pred)) {
              this.diagnostics.push({
                line: 0,
                column: 0,
                message: `Node '${nodeId}': unknown semantic predicate '${pred}'. Valid: ${validNames.join(
                  ', '
                )}`,
                severity: 'warning',
              });
            }
          } else if (typeof pred === 'object' && pred.kind) {
            // Full predicate object -- Phase 8 will handle validation.
          }
        }
      } catch {
        this.diagnostics.push({
          line: 0,
          column: 0,
          message: `Node '${nodeId}': semantic_predicates is not valid JSON`,
          severity: 'warning',
        });
      }
    }
  }

  public getCompletions(line: string, column: number): string[] {
    const keywords = [
      'FORK',
      'RACE',
      'FOLD',
      'VENT',
      'PROCESS',
      'COLLAPSE',
      'TUNNEL',
      'SLIVER',
      'LAMINAR',
      'MEASURE',
      'HALT',
      'EVOLVE',
      'ENTANGLE',
      'SUPERPOSE',
      'OBSERVE',
      'METACOG',
    ];
    const nodeIds = Array.from(this.ast.nodes.keys());

    const prefix =
      line
        .slice(0, column)
        .split(/[^A-Za-z0-9_]+/)
        .pop()
        ?.toUpperCase() || '';
    return [...keywords, ...nodeIds].filter((w) => w.startsWith(prefix));
  }

  public parse(input: string): BettyParseResult {
    const normalizedInput = lowerUfcsSource(input);
    if (!normalizedInput.trim()) {
      const emptyAst: GraphAST = { nodes: new Map(), edges: [] };
      const verification = runAllVerificationPasses(emptyAst, 0);
      return {
        ast: null,
        output: '',
        b1: 0,
        diagnostics: [],
        buleyMeasure: 0,
        stability: null,
        optimizer: null,
        coarseningSynthesis: null,
        voidDimensions: verification.voidDimensions,
        landauerHeat: verification.landauerHeat,
        deficit: null,
        semanticCompatibility: null,
        hopeCertificate: null,
        clockwork: verification.clockwork,
        universe: verification.universe,
        knotBinary: null,
      };
    }

    this.logs = [];
    this.diagnostics = [];
    this.b1 = 0;
    this.ast = { nodes: new Map(), edges: [] };
    this.stability = null;
    this.coarseningSynthesisResult = null;
    this.wasmBridge = new QuantumWasmBridge();

    // 1. REJECT IMPERATIVE CODE (Global check to handle multi-line property blocks)
    // Strip property blocks and comments to avoid false positives.
    const inputWithoutProperties = normalizedInput
      .replace(/\/\/[^\n]*/g, '')
      .replace(/--[^\n]*/g, '')
      .replace(/#[^\n]*/g, '')
      .replace(/\{[^{}]*\}/gs, '{}'); // Use 's' flag for multi-line support
    
    const lines = normalizedInput.split('\n');
    const linesForImperativeCheck = inputWithoutProperties.split('\n');

    for (let i = 0; i < lines.length; i++) {
      const line = (lines[i] ?? '').trim();
      if (!line || line.startsWith('//') || line.startsWith('--') || line.startsWith('#')) continue;

      // Detect standalone source/parameter declarations which add to β₁
      const sourceMatch = line.match(/^\s*\(([^:)\s|]+)\s*:\s*(Source|Parameter)/);
      if (sourceMatch) {
        this.b1++;
      }

      const lineForCheck = (linesForImperativeCheck[i] ?? '').trim();
      const imperativeMatch = lineForCheck.match(IMPERATIVE_KEYWORD_REGEX);
      if (imperativeMatch) {
        this.diagnostics.push({
          line: i + 1,
          column: line.indexOf(imperativeMatch[0]) + 1,
          message: `Imperative keyword '${imperativeMatch[0]}' rejected. Gnosis requires pure topological declarations.`,
          severity: 'error',
        });
        continue;
      }

      // 2. PARSE NODES
      // Only match nodes that are NOT part of an edge definition (followed by -[: or preceded by ]->)
      {
        const parsedNodes = parseNodeDeclarations(line);
        for (const pn of parsedNodes) {
          if (!pn.id) continue;
          const properties = this.parseProperties(pn.propertiesRaw);

          if (!this.ast.nodes.has(pn.id)) {
            this.ast.nodes.set(pn.id, {
              id: pn.id,
              labels: pn.label ? [pn.label] : [],
              properties,
            });
            continue;
          }

          const existing = this.ast.nodes.get(pn.id)!;
          if (pn.label && existing.labels.length === 0) {
            existing.labels = [pn.label];
          }
          if (Object.keys(properties).length > 0) {
            existing.properties = { ...existing.properties, ...properties };
          }
        }
      }

      // 3. PARSE EDGES
      const parsedEdges = parseEdgeDeclarations(line);
      const lineMatched = parsedEdges.length > 0;

      for (const pe of parsedEdges) {
        const edgeType = pe.edgeType;
        const propertiesRaw = pe.propertiesRaw;

        // Parse node refs once, extract IDs from the same pass (no redundant split)
        const sourceRefs = parseNodeRefsFromEdgeGroup(pe.sourceRaw);
        const targetRefs = parseNodeRefsFromEdgeGroup(pe.targetRaw);
        const sources = sourceRefs.map((ref) => ref.id);
        const targets = targetRefs.map((ref) => ref.id);

        // Create/merge nodes from refs in a single pass
        for (const ref of sourceRefs) {
          if (!this.ast.nodes.has(ref.id)) {
            this.ast.nodes.set(ref.id, {
              id: ref.id,
              labels: ref.label ? [ref.label] : [],
              properties: ref.propertiesRaw
                ? this.parseProperties(ref.propertiesRaw)
                : {},
            });
          } else {
            const existing = this.ast.nodes.get(ref.id)!;
            if (ref.label && !existing.labels.includes(ref.label)) {
              existing.labels.push(ref.label);
            }
            if (ref.propertiesRaw) {
              const properties = this.parseProperties(ref.propertiesRaw);
              if (Object.keys(properties).length > 0) {
                existing.properties = { ...existing.properties, ...properties };
              }
            }
          }
        }
        for (const ref of targetRefs) {
          if (!this.ast.nodes.has(ref.id)) {
            this.ast.nodes.set(ref.id, {
              id: ref.id,
              labels: ref.label ? [ref.label] : [],
              properties: ref.propertiesRaw
                ? this.parseProperties(ref.propertiesRaw)
                : {},
            });
          } else {
            const existing = this.ast.nodes.get(ref.id)!;
            if (ref.label && !existing.labels.includes(ref.label)) {
              existing.labels.push(ref.label);
            }
            if (ref.propertiesRaw) {
              const properties = this.parseProperties(ref.propertiesRaw);
              if (Object.keys(properties).length > 0) {
                existing.properties = { ...existing.properties, ...properties };
              }
            }
          }
        }

        // Topological Validation
        if (edgeType === 'FORK') {
          this.b1 += targets.length - 1;
        } else if (
          edgeType === 'FOLD' ||
          edgeType === 'COLLAPSE' ||
          edgeType === 'OBSERVE'
        ) {
          // OBSERVE triggers collapse — reading forces superposition to resolve
          // The topology is append-only (no GC). beta1=0 means converged, not deleted.
          this.b1 = Math.max(0, this.b1 - (sources.length - 1));
        } else if (edgeType === 'RACE' || edgeType === 'SLIVER') {
          this.b1 = Math.max(
            0,
            this.b1 - Math.max(0, sources.length - targets.length)
          );
        } else if (edgeType === 'VENT') {
          this.b1 = Math.max(0, this.b1 - 1);
        } else if (edgeType === 'LAMINAR') {
          // LAMINAR: hella-whipped pipeline — fork codecs internally, race per chunk, fold to smallest.
          // β₁ increases by (codecs - 1) for the internal codec racing, but the pipeline
          // presents as a single stream externally (chunked, compressed, Flow-framed).
          // Net β₁ effect: 0 (internal fork/race/fold is self-contained).
          // The topology sees LAMINAR as a PROCESS-like passthrough that compresses.
        } else if (edgeType === 'METACOG') {
          // METACOG: traced monoidal feedback that reads its own void boundary shape
          // and adapts its parameters. The c0-c3 metacognitive loop:
          //   c0 (execute) → c1 (monitor kurtosis/entropy/inverseBule)
          //   → c2 (evaluate gradient, detect regime changes)
          //   → c3 (adapt eta, exploration rate, gait)
          //   → c0 (execute with adapted parameters)
          //
          // Net β₁ effect: 0 (the feedback loop is self-contained, like LAMINAR).
          // The convergence certificate is the Foster-Lyapunov drift condition:
          // the inverse Bule is the decreasing measure.
          //
          // METACOG edges create cycles in the topology graph, which is normally
          // forbidden. The compiler allows METACOG cycles because the convergence
          // certificate guarantees termination at the Skyrms equilibrium.
          //
          // Trot-canter-gallop gait selection is implicit: the METACOG edge
          // adapts the pipeline depth (1 → 4 → 16) based on kurtosis.
        }

        this.wasmBridge.processAstEdge(
          edgeType,
          sources.length,
          targets.length
        );

        const edgeProperties = this.parseProperties(propertiesRaw);

        this.ast.edges.push({
          sourceIds: sources,
          targetIds: targets,
          type: edgeType,
          properties: edgeProperties,
        });

        // Node creation already handled by sourceRefs/targetRefs above
      }

      if (!lineMatched && !line.startsWith('(')) {
        this.diagnostics.push({
          line: i + 1,
          column: 1,
          message: `Invalid Gnosis syntax. Expected node or edge declaration.`,
          severity: 'info',
        });
      }
    }

    // 4. TOPOLOGICAL INTEGRITY CHECKS
    const referencedNodes = new Set<string>();
    this.ast.edges.forEach((e) => {
      e.sourceIds.forEach((id) => referencedNodes.add(id));
      e.targetIds.forEach((id) => referencedNodes.add(id));
    });

    this.ast.nodes.forEach((node) => {
      if (!referencedNodes.has(node.id)) {
        this.diagnostics.push({
          line: 1,
          column: 1,
          message: `Disconnected node '${node.id}' detected. It will not participate in the covering space.`,
          severity: 'warning',
        });
      }
    });

    this.checkTaggedNodeExhaustiveness();
    this.checkStructuredConcurrencyProperties();
    this.checkVoidWalkerInvariants();
    this.checkFoldEthics();
    this.checkDeficitClosure();

    const injectionResult = injectSensitiveZkEnvelopes(this.ast);
    this.ast = injectionResult.ast;
    if (injectionResult.injected.length > 0) {
      this.diagnostics.push({
        line: 1,
        column: 1,
        message: `Auto-injected ${injectionResult.injected.length} ZK envelope node(s) for sensitive sync/materialization flows.`,
        severity: 'info',
      });
    }

    // Run shared verification passes (stability, optimizer, coarsening, semantic, etc.)
    const verification = runAllVerificationPasses(this.ast, this.b1);
    this.ast = verification.ast;
    this.stability = verification.stability;
    this.optimizerResult = verification.optimizer;
    this.coarseningSynthesisResult = verification.coarseningSynthesis;
    this.semanticResult = verification.semanticCompatibility;
    this.diagnostics.push(...verification.diagnostics);

    // Betty-specific checks that require compiler instance state
    this.checkFirstLaw();
    this.checkBuleyeanPositivity();
    this.checkEntangleDimensions();
    this.checkSelfVerificationAnnotations();
    this.checkAlephCompleteness();
    this.checkEntropyReversal();
    this.validateSemanticPredicates();

    const buleyMeasure = this.getBuleyMeasurement();
    const spectralSummary =
      this.stability?.spectralRadius !== null &&
      this.stability?.spectralRadius !== undefined
        ? `\nSpectral Radius: ${this.stability.spectralRadius.toFixed(3)}`
        : '';
    const optimizerSummary =
      this.optimizerResult && this.optimizerResult.passesApplied.length > 0
        ? `\nOptimizer: ${this.optimizerResult.passesApplied.join(', ')} (${
            this.optimizerResult.certificates.length
          } certificate(s))`
        : '';
    const summary = `[Betty Professional Compiler]\nNodes: ${
      this.ast.nodes.size
    }, Edges: ${this.ast.edges.length}\nBetti: ${
      this.b1
    }, Buley Measure: ${buleyMeasure.toFixed(
      2
    )}${spectralSummary}${optimizerSummary}`;

    return {
      ast: this.ast,
      output: summary,
      b1: this.b1,
      diagnostics: this.diagnostics,
      buleyMeasure,
      stability: this.stability,
      optimizer: this.optimizerResult,
      coarseningSynthesis: this.coarseningSynthesisResult,
      voidDimensions: verification.voidDimensions,
      landauerHeat: verification.landauerHeat,
      deficit: verification.deficit,
      semanticCompatibility: this.semanticResult,
      hopeCertificate: verification.hopeCertificate,
      clockwork: verification.clockwork,
      universe: verification.universe,
      knotBinary: this.emitKnot(),
    };
  }

  public parseLegalese(
    input: string,
    options: LegaleseLoweringOptions = {}
  ): BettyLegaleseParseResult {
    const lowering = lowerLegaleseToGg(input, options);
    return {
      ...this.parse(lowering.ggSource),
      ggSource: lowering.ggSource,
      lowering,
    };
  }

  /**
   * Emit a packed knot binary from the current AST.
   * The knot formalizes the wire format: 2 bits per crossing, zero overhead.
   */
  private emitKnot(): Uint8Array | null {
    if (!this.ast || this.ast.edges.length === 0) return null;
    try {
      const { coil, toBytes } =
        // eslint-disable-next-line @typescript-eslint/no-require-imports
        require('../runtime/packed-knot.js') as typeof import('../runtime/packed-knot.js');
      return toBytes(coil(this.ast));
    } catch {
      return null;
    }
  }

  private parseProperties(propertiesRaw: string): Record<string, string> {
    return parsePropertiesShared(propertiesRaw);
  }

  private checkTaggedNodeExhaustiveness(): void {
    for (const node of this.ast.nodes.values()) {
      const taggedDefinition = this.resolveTaggedNodeDefinition(node);
      if (!taggedDefinition) {
        continue;
      }

      const { label: taggedLabel, expectedCases } = taggedDefinition;
      const outgoingEdges = this.ast.edges.filter((edge) =>
        edge.sourceIds.some((sourceId) => sourceId.trim() === node.id)
      );
      if (outgoingEdges.length === 0) {
        continue;
      }

      const caseEdges = outgoingEdges
        .map((edge) => ({
          edge,
          cases: this.extractTaggedCases(edge),
        }))
        .filter((entry) => entry.cases.length > 0);

      if (caseEdges.length === 0) {
        continue;
      }

      if (expectedCases.length === 0) {
        this.diagnostics.push({
          line: 1,
          column: 1,
          message: `${taggedLabel} node '${node.id}' declares tagged exits but no closed cases. Add a cases property to enable exhaustiveness checks.`,
          severity: 'warning',
        });
        continue;
      }

      const unconstrainedEdgeCount = outgoingEdges.length - caseEdges.length;
      if (unconstrainedEdgeCount > 0) {
        this.diagnostics.push({
          line: 1,
          column: 1,
          message: `${taggedLabel} node '${node.id}' mixes tagged case edges with unconstrained exits; exhaustiveness cannot be guaranteed.`,
          severity: 'warning',
        });
      }

      const seenCases = new Map<string, number>();
      for (const { edge, cases } of caseEdges) {
        for (const edgeCase of cases) {
          if (!expectedCases.includes(edgeCase)) {
            this.diagnostics.push({
              line: 1,
              column: 1,
              message: `${taggedLabel} node '${
                node.id
              }' uses unknown case '${edgeCase}' on ${
                edge.type
              }. Expected one of: ${expectedCases.join(', ')}.`,
              severity: 'error',
            });
            continue;
          }

          const existingCount = seenCases.get(edgeCase) ?? 0;
          seenCases.set(edgeCase, existingCount + 1);
          if (existingCount > 0) {
            this.diagnostics.push({
              line: 1,
              column: 1,
              message: `${taggedLabel} node '${node.id}' routes case '${edgeCase}' more than once.`,
              severity: 'error',
            });
          }
        }
      }

      const missingCases = expectedCases.filter(
        (edgeCase) => !seenCases.has(edgeCase)
      );
      if (missingCases.length > 0) {
        this.diagnostics.push({
          line: 1,
          column: 1,
          message: `${taggedLabel} node '${
            node.id
          }' is missing tagged routes for: ${missingCases.join(', ')}.`,
          severity: 'error',
        });
      }
    }
  }

  private extractTaggedCases(edge: ASTEdge): string[] {
    const rawCase =
      edge.properties.case ??
      edge.properties.match ??
      edge.properties.when ??
      edge.properties.variant ??
      edge.properties.kind ??
      edge.properties.status;
    if (!rawCase) {
      return [];
    }

    return rawCase
      .split(/[\s,|]+/)
      .map((entry) => entry.trim().toLowerCase())
      .filter((entry) => entry.length > 0);
  }

  private resolveTaggedNodeDefinition(
    node: ASTNode
  ): { label: string; expectedCases: string[] } | null {
    const nativeTaggedLabel = node.labels.find(
      (label) => NATIVE_TAGGED_NODE_CASES[label]
    );
    if (nativeTaggedLabel) {
      return {
        label: nativeTaggedLabel,
        expectedCases: [...(NATIVE_TAGGED_NODE_CASES[nativeTaggedLabel] ?? [])],
      };
    }

    if (node.labels.includes('Variant')) {
      return {
        label: 'Variant',
        expectedCases: this.parseClosedCases(
          node.properties.cases ??
            node.properties.variants ??
            node.properties.options
        ),
      };
    }

    return null;
  }

  private parseClosedCases(rawCases: string | undefined): string[] {
    if (!rawCases) {
      return [];
    }

    return [
      ...new Set(
        rawCases
          .split(/[\s,|]+/)
          .map((entry) => entry.trim().toLowerCase())
          .filter((entry) => entry.length > 0)
      ),
    ];
  }

  private checkStructuredConcurrencyProperties(): void {
    for (const edge of this.ast.edges) {
      if (
        edge.type !== 'RACE' &&
        edge.type !== 'FOLD' &&
        edge.type !== 'COLLAPSE'
      ) {
        continue;
      }

      const rawFailure = edge.properties.failure?.trim().toLowerCase();
      if (rawFailure && !STRUCTURED_CONCURRENCY_FAILURES.has(rawFailure)) {
        this.diagnostics.push({
          line: 1,
          column: 1,
          message: `${edge.type} uses unknown failure policy '${rawFailure}'. Expected one of: cancel, vent, shield.`,
          severity: 'error',
        });
      }

      for (const field of ['timeoutMs', 'deadlineMs', 'timeout', 'deadline']) {
        const rawValue = edge.properties[field];
        if (!rawValue) {
          continue;
        }

        const parsed = Number(rawValue);
        if (!Number.isFinite(parsed) || parsed < 0) {
          this.diagnostics.push({
            line: 1,
            column: 1,
            message: `${edge.type} requires '${field}' to be a non-negative number.`,
            severity: 'error',
          });
        }
      }
    }
  }

  // ── Void Walker Static Lint ──────────────────────────────────────
  //
  // The 8 negotiation invariants, checked at compile time.
  //
  // METACOG is the first-class edge for void walker feedback loops.
  // It carries an implicit Foster-Lyapunov convergence certificate
  // (the inverse Bule is the decreasing measure) and implicit traced
  // monoidal structure (vanishing + yanking satisfied by the c0-c3
  // loop's Skyrms equilibrium guarantee).
  //
  // PROCESS edges that form cycles are the *uncertified* path --
  // the author must supply convergence and axiom annotations manually.
  //
  // All 8 rules are warnings by default, errors with strict: "true".

  private checkVoidWalkerInvariants(): void {
    const strict = this.isStrictMode();
    const severity: 'error' | 'warning' = strict ? 'error' : 'warning';

    this.checkEtaBounded(severity);
    this.checkExplorationBounded(severity);
    this.checkComplementDistributionValid(severity);
    this.checkVoidBoundaryMonotone(severity);
    this.checkGaitTransitionsValid(severity);
    this.checkMetacogFeedbackLoop(severity);
    this.checkConservation(severity);
    this.checkTracedMonoidal(severity);
  }

  private isStrictMode(): boolean {
    for (const node of this.ast.nodes.values()) {
      if (node.properties.strict === 'true') return true;
    }
    return false;
  }

  /** 1. eta_bounded: eta in [0.5, 10.0] on nodes and METACOG edge properties */
  private checkEtaBounded(severity: 'error' | 'warning'): void {
    for (const node of this.ast.nodes.values()) {
      const raw = node.properties.eta;
      if (raw === undefined) continue;

      const eta = Number(raw);
      if (!Number.isFinite(eta) || eta < 0.5 || eta > 10.0) {
        this.diagnostics.push({
          line: 1,
          column: 1,
          code: 'VOID_ETA_UNBOUNDED',
          message: `Node '${node.id}' has eta=${raw} outside allowed range [0.5, 10.0].`,
          severity,
        });
      }
    }

    // METACOG edges may carry eta as an adaptation target -- still must be bounded
    for (const edge of this.ast.edges) {
      if (edge.type !== 'METACOG') continue;
      const raw = edge.properties.eta;
      if (raw === undefined) continue;

      const eta = Number(raw);
      if (!Number.isFinite(eta) || eta < 0.5 || eta > 10.0) {
        this.diagnostics.push({
          line: 1,
          column: 1,
          code: 'VOID_METACOG_PARAMETER_DRIFT',
          message: `METACOG edge carries eta=${raw} outside allowed range [0.5, 10.0]. The c3 adaptation target must respect the same bounds as nodes.`,
          severity,
        });
      }
    }
  }

  /** 2. exploration_bounded: exploration in [0.01, 0.5] on nodes and METACOG edges */
  private checkExplorationBounded(severity: 'error' | 'warning'): void {
    for (const node of this.ast.nodes.values()) {
      const raw = node.properties.exploration;
      if (raw === undefined) continue;

      const exploration = Number(raw);
      if (
        !Number.isFinite(exploration) ||
        exploration < 0.01 ||
        exploration > 0.5
      ) {
        this.diagnostics.push({
          line: 1,
          column: 1,
          code: 'VOID_EXPLORATION_UNBOUNDED',
          message: `Node '${node.id}' has exploration=${raw} outside allowed range [0.01, 0.5].`,
          severity,
        });
      }
    }

    for (const edge of this.ast.edges) {
      if (edge.type !== 'METACOG') continue;
      const raw = edge.properties.exploration;
      if (raw === undefined) continue;

      const exploration = Number(raw);
      if (
        !Number.isFinite(exploration) ||
        exploration < 0.01 ||
        exploration > 0.5
      ) {
        this.diagnostics.push({
          line: 1,
          column: 1,
          code: 'VOID_METACOG_PARAMETER_DRIFT',
          message: `METACOG edge carries exploration=${raw} outside allowed range [0.01, 0.5].`,
          severity,
        });
      }
    }
  }

  /** 3. complement_distribution_valid: any edge with a distribution must sum to 1.0 */
  private checkComplementDistributionValid(
    severity: 'error' | 'warning'
  ): void {
    for (const edge of this.ast.edges) {
      const raw = edge.properties.distribution;
      if (raw === undefined) continue;

      const values = raw
        .replace(/^\[|\]$/g, '')
        .split(/[\s,]+/)
        .map(Number)
        .filter(Number.isFinite);

      if (values.length === 0) continue;

      const sum = values.reduce((a, b) => a + b, 0);
      if (Math.abs(sum - 1.0) > 1e-9) {
        this.diagnostics.push({
          line: 1,
          column: 1,
          code: 'VOID_COMPLEMENT_DISTRIBUTION_INVALID',
          message: `${edge.type} edge from [${edge.sourceIds.join(
            ', '
          )}] has distribution summing to ${sum.toFixed(6)}, expected 1.0.`,
          severity,
        });
      }
    }
  }

  /** 4. void_boundary_monotone: boundary counts only increase along PROCESS and METACOG edges */
  private checkVoidBoundaryMonotone(severity: 'error' | 'warning'): void {
    for (const edge of this.ast.edges) {
      if (edge.type !== 'PROCESS' && edge.type !== 'METACOG') continue;

      for (const sourceId of edge.sourceIds) {
        const sourceNode = this.ast.nodes.get(sourceId);
        if (!sourceNode?.properties.void_boundary) continue;

        for (const targetId of edge.targetIds) {
          const targetNode = this.ast.nodes.get(targetId);
          if (!targetNode?.properties.void_boundary) continue;

          const sourceBound = Number(sourceNode.properties.void_boundary);
          const targetBound = Number(targetNode.properties.void_boundary);

          if (
            Number.isFinite(sourceBound) &&
            Number.isFinite(targetBound) &&
            targetBound < sourceBound
          ) {
            this.diagnostics.push({
              line: 1,
              column: 1,
              code: 'VOID_BOUNDARY_NON_MONOTONE',
              message: `Void boundary decreases from '${sourceId}' (${sourceBound}) to '${targetId}.js' (${targetBound}). Boundary counts can only increase.`,
              severity,
            });
          }
        }
      }
    }
  }

  /** 5. gait_transitions_valid: stand->trot->canter->gallop or downshift one level */
  private checkGaitTransitionsValid(severity: 'error' | 'warning'): void {
    for (const edge of this.ast.edges) {
      // Gait transitions matter on PROCESS and METACOG edges.
      // METACOG adapts gait implicitly via kurtosis, but the adapted
      // gait still must obey the transition lattice.
      for (const sourceId of edge.sourceIds) {
        const sourceNode = this.ast.nodes.get(sourceId);
        if (!sourceNode?.properties.gait) continue;

        for (const targetId of edge.targetIds) {
          const targetNode = this.ast.nodes.get(targetId);
          if (!targetNode?.properties.gait) continue;

          const srcIdx = GAIT_ORDER.indexOf(sourceNode.properties.gait);
          const tgtIdx = GAIT_ORDER.indexOf(targetNode.properties.gait);

          if (srcIdx < 0 || tgtIdx < 0) continue;

          if (tgtIdx - srcIdx < -1) {
            this.diagnostics.push({
              line: 1,
              column: 1,
              code: 'VOID_GAIT_TRANSITION_INVALID',
              message: `Invalid gait transition from '${sourceNode.properties.gait}' to '${targetNode.properties.gait}' at '${sourceId}'->'${targetId}.js'. Can only upshift or downshift one level.`,
              severity,
            });
          }
        }
      }
    }
  }

  /**
   * 6. metacog_feedback_loop: PROCESS cycles need explicit convergence predicates.
   *    METACOG edges are exempt -- their convergence certificate is the
   *    Foster-Lyapunov drift condition (inverse Bule as decreasing measure).
   */
  private checkMetacogFeedbackLoop(severity: 'error' | 'warning'): void {
    // Build adjacency for cycle-forming edge types
    const adj = new Map<string, ASTEdge[]>();
    for (const edge of this.ast.edges) {
      if (edge.type !== 'PROCESS' && edge.type !== 'METACOG') continue;
      for (const src of edge.sourceIds) {
        if (!adj.has(src)) adj.set(src, []);
        adj.get(src)!.push(edge);
      }
    }

    const visited = new Set<string>();
    const stack = new Set<string>();
    const cycleEdges: ASTEdge[] = [];

    const dfs = (nodeId: string): void => {
      visited.add(nodeId);
      stack.add(nodeId);

      for (const edge of adj.get(nodeId) ?? []) {
        for (const target of edge.targetIds) {
          if (stack.has(target)) {
            cycleEdges.push(edge);
          } else if (!visited.has(target)) {
            dfs(target);
          }
        }
      }

      stack.delete(nodeId);
    };

    for (const nodeId of adj.keys()) {
      if (!visited.has(nodeId)) dfs(nodeId);
    }

    for (const edge of cycleEdges) {
      // METACOG edges carry implicit convergence -- no annotation needed
      if (edge.type === 'METACOG') continue;

      if (!edge.properties.convergence) {
        this.diagnostics.push({
          line: 1,
          column: 1,
          code: 'VOID_METACOG_MISSING_CONVERGENCE',
          message: `PROCESS edge from [${edge.sourceIds.join(
            ', '
          )}] to [${edge.targetIds.join(
            ', '
          )}] creates a cycle but has no convergence predicate. Use a METACOG edge for implicit convergence certification.`,
          severity,
        });
      }
    }
  }

  /** 7. conservation: fork output count = fold input count (no paths lost or created) */
  private checkConservation(severity: 'error' | 'warning'): void {
    for (const edge of this.ast.edges) {
      if (edge.type !== 'FOLD') {
        continue;
      }

      const foldSources = edge.sourceIds.map((sourceId) => sourceId.trim());
      const coveringFork = this.ast.edges.find((candidate) => {
        if (candidate.type !== 'FORK') {
          return false;
        }

        const forkTargets = candidate.targetIds.map((targetId) =>
          targetId.trim()
        );
        return foldSources.every((sourceId) => forkTargets.includes(sourceId));
      });

      if (!coveringFork) {
        continue;
      }

      const forkWidth = coveringFork.targetIds.length;
      const foldWidth = foldSources.length;
      if (forkWidth === foldWidth) {
        continue;
      }

      this.diagnostics.push({
        line: 1,
        column: 1,
        code: 'VOID_CONSERVATION_VIOLATED',
        message: `Conservation violated: FORK outputs ${forkWidth} paths but FOLD inputs ${foldWidth}. Fork output count must equal fold input count.`,
        severity,
      });
    }
  }

  /**
   * 8. traced_monoidal: feedback loops must satisfy vanishing + yanking axioms.
   *    METACOG self-loops are exempt -- the c0-c3 loop satisfies both axioms
   *    via the Skyrms equilibrium guarantee (vanishing: inverse Bule converges
   *    to zero at equilibrium; yanking: removing the feedback wire yields the
   *    same steady-state distribution).
   */
  private checkTracedMonoidal(severity: 'error' | 'warning'): void {
    for (const edge of this.ast.edges) {
      const selfLoop = edge.sourceIds.some((s) => edge.targetIds.includes(s));
      if (!selfLoop) continue;

      // METACOG carries implicit traced monoidal structure
      if (edge.type === 'METACOG') continue;

      const hasVanishing = edge.properties.vanishing !== undefined;
      const hasYanking = edge.properties.yanking !== undefined;

      if (!hasVanishing || !hasYanking) {
        const missing = [
          !hasVanishing ? 'vanishing' : null,
          !hasYanking ? 'yanking' : null,
        ]
          .filter(Boolean)
          .join(', ');

        this.diagnostics.push({
          line: 1,
          column: 1,
          code: 'VOID_TRACED_MONOIDAL_VIOLATED',
          message: `Feedback loop on [${edge.sourceIds.join(
            ', '
          )}] is missing traced monoidal axiom annotations: ${missing}. Use a METACOG edge for implicit axiom satisfaction.`,
          severity,
        });
      }
    }
  }

  // ============================================================================
  // Fold Ethics Checker -- static lint for the 5x5 ethics grid
  //
  // The fold ethics grid (§14.5.7) maps each primitive to an ethical
  // operation at each void condition. The checker validates that
  // topologies respect the ethical properties of their primitives:
  //
  //   FORK creates possibility, destroys nothing.
  //   RACE creates knowledge, destroys nothing.
  //   FOLD creates commitment, destroys alternatives.
  //   VENT creates clarity, destroys nuance.
  //   TRACE creates learning, destroys ignorance.
  //
  // Violations are warnings by default, errors with ethics: "strict".
  // ============================================================================

  private checkFoldEthics(): void {
    const strict = this.isEthicsStrict();
    const severity: 'error' | 'warning' = strict ? 'error' : 'warning';

    this.checkFoldWithoutEvidence(severity);
    this.checkFoldDestroysAll(severity);
    this.checkVentWithoutBoundary(severity);
    this.checkRacePrematureCollapse(severity);
    this.checkForkZeroOptions(severity);
    this.checkMissingVentPath(severity);
    this.checkIrreversibleOnOthersVoid(severity);
    this.checkNoTrace(severity);
  }

  private isEthicsStrict(): boolean {
    for (const node of this.ast.nodes.values()) {
      if (node.properties.ethics === 'strict') return true;
    }
    return false;
  }

  /**
   * 1. FOLD without evidence: folding on sparse void is Courage, but folding
   *    on *empty* void is reckless. A FOLD/RACE/COLLAPSE edge whose source
   *    nodes have no void_boundary annotation and no upstream FORK/RACE
   *    producing tombstones is a blind commitment.
   */
  private checkFoldWithoutEvidence(severity: 'error' | 'warning'): void {
    const foldEdges = this.ast.edges.filter(
      (e) => e.type === 'FOLD' || e.type === 'COLLAPSE'
    );

    for (const edge of foldEdges) {
      // Check if any source node has void evidence
      const hasEvidence = edge.sourceIds.some((srcId) => {
        return this.hasUpstreamVoidEvidence(srcId.trim(), new Set());
      });

      if (!hasEvidence) {
        this.diagnostics.push({
          line: 1,
          column: 1,
          code: 'ETHICS_FOLD_WITHOUT_EVIDENCE',
          message: `FOLD from [${edge.sourceIds.join(
            ', '
          )}] has no upstream void evidence. Folding without evidence is a blind commitment -- the grid cell [Fold, Sparse void] is Courage, but [Fold, Empty void] has no ethical name. Add void_density or an upstream RACE/FORK.`,
          severity,
        });
      }
    }
  }

  /**
   * 2. FOLD destroys all: a FOLD/COLLAPSE whose target count is zero
   *    or whose source count equals the total node count minus one.
   *    A fold that eliminates ALL alternatives leaves no future forks --
   *    this is ethically distinct from Judgment (which preserves at least
   *    the complement of what was folded).
   */
  private checkFoldDestroysAll(severity: 'error' | 'warning'): void {
    for (const edge of this.ast.edges) {
      if (edge.type !== 'FOLD' && edge.type !== 'COLLAPSE') continue;

      // If the fold consumes all nodes in the topology, flag it
      const totalNodes = this.ast.nodes.size;
      if (edge.sourceIds.length >= totalNodes - 1 && totalNodes > 2) {
        this.diagnostics.push({
          line: 1,
          column: 1,
          code: 'ETHICS_FOLD_DESTROYS_ALL',
          message: `FOLD from [${edge.sourceIds.join(', ')}] consumes ${
            edge.sourceIds.length
          } of ${totalNodes} nodes. A fold that destroys all alternatives is not Judgment -- it is annihilation. Preserve at least one alternative path.`,
          severity,
        });
      }
    }
  }

  /**
   * 3. VENT without boundary: venting creates clarity only when it defines
   *    a region. A VENT edge with no condition property is unbounded rejection --
   *    it vents without saying what it's protecting. Boundaries (grid cell
   *    [Vent, Dense void]) require the vent to define what stays.
   */
  private checkVentWithoutBoundary(severity: 'error' | 'warning'): void {
    for (const edge of this.ast.edges) {
      if (edge.type !== 'VENT') continue;

      const hasCondition = edge.properties.condition !== undefined;
      const hasReason = edge.properties.reason !== undefined;

      if (!hasCondition && !hasReason) {
        this.diagnostics.push({
          line: 1,
          column: 1,
          code: 'ETHICS_VENT_WITHOUT_BOUNDARY',
          message: `VENT from [${edge.sourceIds.join(
            ', '
          )}] has no condition or reason. Rejection without stated boundary is not Honesty -- it is arbitrary exclusion. Add condition or reason to define what the vent protects.`,
          severity,
        });
      }
    }
  }

  /**
   * 4. RACE premature collapse: a RACE edge whose sources have no upstream
   *    FORK is racing zero paths -- or racing a single path against nothing.
   *    Patience (grid cell [Race, Dense void]) requires multiple paths to
   *    have been tried before the race resolves.
   */
  private checkRacePrematureCollapse(severity: 'error' | 'warning'): void {
    for (const edge of this.ast.edges) {
      if (edge.type !== 'RACE') continue;

      if (edge.sourceIds.length < 2) {
        this.diagnostics.push({
          line: 1,
          column: 1,
          code: 'ETHICS_RACE_PREMATURE_COLLAPSE',
          message: `RACE from [${edge.sourceIds.join(', ')}] has ${
            edge.sourceIds.length
          } source(s). Racing fewer than two paths is not Listening -- it is premature collapse. A race requires at least two paths to be fair.`,
          severity,
        });
      }
    }
  }

  /**
   * 5. FORK zero options: a FORK edge with zero or one target is not
   *    creating possibility -- it is a disguised PROCESS. Generosity
   *    (grid cell [Fork, Sparse void]) requires at least two paths.
   */
  private checkForkZeroOptions(severity: 'error' | 'warning'): void {
    for (const edge of this.ast.edges) {
      if (edge.type !== 'FORK') continue;

      if (edge.targetIds.length < 2) {
        this.diagnostics.push({
          line: 1,
          column: 1,
          code: 'ETHICS_FORK_ZERO_OPTIONS',
          message: `FORK from [${edge.sourceIds.join(', ')}] creates ${
            edge.targetIds.length
          } path(s). A fork that creates fewer than two paths is not Generosity -- it is a disguised PROCESS. Use PROCESS if only one path is intended.`,
          severity,
        });
      }
    }
  }

  /**
   * 6. Missing vent path: a topology with FORK and FOLD but no VENT
   *    has no way to reject. Without rejection, there is no Honesty,
   *    no Boundaries, no Criticism. The void boundary cannot grow.
   *    This is ethically suspect: a system that cannot say no cannot learn.
   */
  private checkMissingVentPath(severity: 'error' | 'warning'): void {
    const hasFork = this.ast.edges.some((e) => e.type === 'FORK');
    const hasFold = this.ast.edges.some(
      (e) => e.type === 'FOLD' || e.type === 'RACE' || e.type === 'COLLAPSE'
    );
    const hasVent = this.ast.edges.some((e) => e.type === 'VENT');
    const hasTunnel = this.ast.edges.some((e) => e.type === 'TUNNEL');
    const hasImplicitVent = this.ast.edges.some((e) =>
      IMPLICIT_VENT_EDGE_TYPES.has(e.type)
    );
    const preservesMultiplicity = this.ast.edges.some(
      (e) =>
        e.type === 'FOLD' &&
        e.properties.strategy?.trim().toLowerCase() === 'preserve-multiplicity'
    );

    if (
      hasFork &&
      hasFold &&
      !hasVent &&
      !hasTunnel &&
      !hasImplicitVent &&
      !preservesMultiplicity
    ) {
      this.diagnostics.push({
        line: 1,
        column: 1,
        code: 'ETHICS_MISSING_VENT_PATH',
        message: `Topology has FORK and FOLD but no VENT or TUNNEL. A system that cannot reject cannot learn from failure. Add a VENT path to enable Boundaries, Honesty, and void boundary growth.`,
        severity,
      });
    }
  }

  /**
   * 7. Irreversible operation on other's void: a FOLD or VENT edge
   *    annotated with void_source: "other" is operating on someone else's
   *    rejection history. Grid cell [Fold, Other's void] = Sacrifice,
   *    and [Vent, Other's void] = Tough love. Both are valid but require
   *    explicit acknowledgment (sacrifice: "true" or tough_love: "true").
   */
  private checkIrreversibleOnOthersVoid(severity: 'error' | 'warning'): void {
    for (const edge of this.ast.edges) {
      if (
        edge.type !== 'FOLD' &&
        edge.type !== 'VENT' &&
        edge.type !== 'COLLAPSE'
      )
        continue;
      if (edge.properties.void_source !== 'other') continue;

      const isSacrifice = edge.properties.sacrifice === 'true';
      const isToughLove = edge.properties.tough_love === 'true';

      if (!isSacrifice && !isToughLove) {
        const op = edge.type === 'VENT' ? 'Tough love' : 'Sacrifice';
        this.diagnostics.push({
          line: 1,
          column: 1,
          code: 'ETHICS_IRREVERSIBLE_ON_OTHERS_VOID',
          message: `${edge.type} from [${edge.sourceIds.join(
            ', '
          )}] operates on void_source: "other" without explicit acknowledgment. This is ${op} -- it requires ${
            edge.type === 'VENT' ? 'tough_love' : 'sacrifice'
          }: "true" to confirm intent.`,
          severity,
        });
      }
    }
  }

  /**
   * 8. No trace: a topology with FOLD or VENT but no feedback loop
   *    (no cycle, no METACOG, no MEASURE) has no mechanism for learning.
   *    Without Trace there is no Growth, no Dialogue, no Culture.
   *    The void accumulates but is never read back.
   */
  private checkNoTrace(severity: 'error' | 'warning'): void {
    const hasIrreversible = this.ast.edges.some(
      (e) =>
        e.type === 'FOLD' ||
        e.type === 'VENT' ||
        e.type === 'RACE' ||
        e.type === 'COLLAPSE'
    );
    if (!hasIrreversible) return;

    const hasTraceSupportingEdge = this.ast.edges.some((e) =>
      TRACE_SUPPORTING_EDGE_TYPES.has(e.type)
    );

    // Check for any feedback cycle (PROCESS edge back to an earlier node)
    const hasCycle = this.ast.edges.some((edge) =>
      edge.sourceIds.some((s) => edge.targetIds.includes(s))
    );

    // Check for trace-like labels on any node
    const hasTraceNode = Array.from(this.ast.nodes.values()).some((n) =>
      n.labels.some(
        (l) =>
          l.toLowerCase().includes('trace') ||
          l.toLowerCase().includes('measure') ||
          l.toLowerCase().includes('reflect')
      )
    );

    if (!hasTraceSupportingEdge && !hasCycle && !hasTraceNode) {
      this.diagnostics.push({
        line: 1,
        column: 1,
        code: 'ETHICS_NO_TRACE',
        message: `Topology has irreversible operations (FOLD/VENT/RACE) but no trace mechanism (MEASURE, METACOG, feedback cycle, or Trace/Reflect node). A system that folds without learning has no path to Growth or Redemption.`,
        severity,
      });
    }
  }

  // ============================================================================
  // THM-SERVER-OPTIMALITY Paradigm Enforcement: Zero-Deficit Closure
  //
  // The formal surface (THM-ZERO-DEFICIT-PRESERVES-INFORMATION) proves that
  // zero deficit at every layer boundary enables lossless information transport.
  // Positive deficit at a sink node means unclosed forks -- the topology
  // cannot achieve the optimal throughput bound (THM-ROTATION-MAKESPAN-BOUND).
  // ============================================================================

  private checkDeficitClosure(): void {
    if (this.b1 === 0) {
      return; // Zero deficit at final boundary -- paradigm satisfied
    }

    // Find sink nodes
    const sinkNodes = [...this.ast.nodes.values()].filter((node) =>
      node.labels.some(
        (label) =>
          label.toUpperCase() === 'SINK' || label.toUpperCase() === 'TERMINAL'
      )
    );

    // If there are sink nodes declared and beta1 > 0, that's a deficit violation
    if (sinkNodes.length > 0) {
      this.diagnostics.push({
        line: 1,
        column: 1,
        code: 'ERR_DEFICIT_NONZERO',
        message: `Topological deficit violation: β₁=${this.b1} at sink boundary (expected 0). ${this.b1} unclosed fork path(s) remain. By THM-ZERO-DEFICIT-PRESERVES-INFORMATION, positive deficit forces information loss. By THM-ROTATION-MAKESPAN-BOUND, the optimal throughput bound requires zero deficit at every layer boundary. Add FOLD, RACE, or VENT edges to close all fork paths before the sink.`,
        severity: 'error',
      });
    } else if (this.b1 > 0) {
      // No explicit sink but beta1 > 0 -- warn about unclosed topology
      this.diagnostics.push({
        line: 1,
        column: 1,
        code: 'ERR_DEFICIT_NONZERO',
        message: `Topology has final β₁=${this.b1} (${this.b1} unclosed fork path(s)). For optimal throughput (THM-SERVER-OPTIMALITY), all fork paths should collapse to β₁=0 at the topology boundary.`,
        severity: 'warning',
      });
    }
  }

  // ============================================================================
  // Phase 2: Landauer Heat and First Law
  // ============================================================================

  private computeLandauerHeat(): number {
    let heat = 0;
    for (const edge of this.ast.edges) {
      if (
        edge.type === 'FOLD' ||
        edge.type === 'COLLAPSE' ||
        edge.type === 'OBSERVE'
      ) {
        heat += Math.log2(Math.max(2, edge.sourceIds.length));
      } else if (edge.type === 'VENT') {
        heat += 1;
      }
    }
    return heat;
  }

  private checkFirstLaw(): void {
    // Fork entropy = sum of log2(targets) for each FORK
    let forkEntropy = 0;
    for (const edge of this.ast.edges) {
      if (
        edge.type === 'FORK' ||
        edge.type === 'EVOLVE' ||
        edge.type === 'SUPERPOSE'
      ) {
        forkEntropy += Math.log2(Math.max(1, edge.targetIds.length));
      }
    }

    // Erasure = Landauer heat (fold + vent)
    const erasure = this.computeLandauerHeat();

    // First law: fork entropy >= erasure (conservation of information)
    if (forkEntropy > 0 && erasure > 0 && forkEntropy < erasure - 1e-9) {
      this.diagnostics.push({
        line: 1,
        column: 1,
        code: 'THERMO_FIRST_LAW_VIOLATED',
        message: `First law violated: fork entropy (${forkEntropy.toFixed(
          3
        )}) < erasure cost (${erasure.toFixed(
          3
        )}). More information is being erased than was created. Add FORK paths or remove FOLD/VENT edges.`,
        severity: 'warning',
      });
    }
  }

  // ============================================================================
  // Phase 3: Buleyean Positivity
  // ============================================================================

  private checkBuleyeanPositivity(): void {
    for (const edge of this.ast.edges) {
      if (edge.type !== 'FOLD' && edge.type !== 'COLLAPSE') continue;

      // Check bule property on nodes
      for (const sourceId of edge.sourceIds) {
        const node = this.ast.nodes.get(sourceId.trim());
        if (!node) continue;
        const buleRaw = node.properties.bule;
        if (buleRaw === undefined) continue;
        const bule = Number(buleRaw);
        if (Number.isFinite(bule) && bule <= 0) {
          this.diagnostics.push({
            line: 1,
            column: 1,
            code: 'BULE_POSITIVITY_VIOLATED',
            message: `Node '${node.id}' has bule=${buleRaw} (non-positive). Buleyean positivity requires all convergence degrees > 0. The sliver guarantee ensures no fold produces probability zero.`,
            severity: 'error',
          });
        }
      }
    }

    // Check exhaustiveness for tagged types -- unvented variants violate the sliver
    for (const node of this.ast.nodes.values()) {
      const taggedDef = this.resolveTaggedNodeDefinition(node);
      if (!taggedDef || taggedDef.expectedCases.length === 0) continue;

      const outgoingEdges = this.ast.edges.filter((e) =>
        e.sourceIds.some((s) => s.trim() === node.id)
      );
      const ventedCases = new Set<string>();
      const foldedCases = new Set<string>();
      for (const e of outgoingEdges) {
        const cases = this.extractTaggedCases(e);
        if (e.type === 'VENT') cases.forEach((c) => ventedCases.add(c));
        if (e.type === 'FOLD' || e.type === 'COLLAPSE')
          cases.forEach((c) => foldedCases.add(c));
      }

      // Unvented variants that also aren't folded means the sliver is broken
      const unhandled = taggedDef.expectedCases.filter(
        (c) => !ventedCases.has(c) && !foldedCases.has(c)
      );
      if (unhandled.length > 0 && outgoingEdges.length > 0) {
        // Only warn if there are case-aware edges (otherwise exhaustiveness checker handles it)
        const hasCaseEdges = outgoingEdges.some(
          (e) => this.extractTaggedCases(e).length > 0
        );
        if (hasCaseEdges) {
          this.diagnostics.push({
            line: 1,
            column: 1,
            code: 'BULE_EXHAUSTIVENESS_INCOMPLETE',
            message: `${taggedDef.label} node '${
              node.id
            }' has unhandled variants: ${unhandled.join(
              ', '
            )}. Unvented variants violate the Buleyean sliver guarantee.`,
            severity: 'warning',
          });
        }
      }
    }
  }

  // ============================================================================
  // Phase 5: Entanglement Dimension Check
  // ============================================================================

  private checkEntangleDimensions(): void {
    for (const edge of this.ast.edges) {
      if (edge.type !== 'ENTANGLE') continue;

      // Check that entangled nodes have compatible dimensionality
      const sourceDims = new Set<number>();
      const targetDims = new Set<number>();

      for (const s of edge.sourceIds) {
        const node = this.ast.nodes.get(s.trim());
        if (node?.properties.dimensions) {
          sourceDims.add(Number(node.properties.dimensions));
        }
      }
      for (const t of edge.targetIds) {
        const node = this.ast.nodes.get(t.trim());
        if (node?.properties.dimensions) {
          targetDims.add(Number(node.properties.dimensions));
        }
      }

      // If both sides declare dimensions, they must match
      if (sourceDims.size > 0 && targetDims.size > 0) {
        const allDims = new Set([...sourceDims, ...targetDims]);
        if (allDims.size > 1) {
          this.diagnostics.push({
            line: 1,
            column: 1,
            code: 'ENTANGLE_DIM_MISMATCH',
            message: `ENTANGLE from [${edge.sourceIds.join(
              ', '
            )}] to [${edge.targetIds.join(
              ', '
            )}] links nodes with different void dimensionalities: ${[
              ...allDims,
            ].join(', ')}. Entangled boundaries must have the same dimensions.`,
            severity: 'error',
          });
        }
      }
    }
  }

  // ============================================================================
  // Phase 6: Self-Verification Annotations
  // ============================================================================

  private checkSelfVerificationAnnotations(): void {
    const validVerify = new Set([
      'positivity',
      'convergence',
      'conservation',
      'deficit_zero',
      'first_law',
    ]);

    for (const edge of this.ast.edges) {
      const verify = edge.properties.verify;
      if (!verify) continue;

      if (!validVerify.has(verify)) {
        this.diagnostics.push({
          line: 1,
          column: 1,
          message: `Unknown verify annotation '${verify}' on ${
            edge.type
          } edge. Expected one of: ${[...validVerify].join(', ')}.`,
          severity: 'error',
        });
        continue;
      }

      // Route to appropriate check
      if (verify === 'positivity') {
        // Check that all source branches have positive Buleyean weight
        for (const sourceId of edge.sourceIds) {
          const node = this.ast.nodes.get(sourceId.trim());
          if (node?.properties.bule) {
            const bule = Number(node.properties.bule);
            if (Number.isFinite(bule) && bule <= 0) {
              this.diagnostics.push({
                line: 1,
                column: 1,
                code: 'BULE_POSITIVITY_VIOLATED',
                message: `verify: 'positivity' failed: node '${node.id}' has bule=${bule} <= 0.`,
                severity: 'error',
              });
            }
          }
        }
      } else if (verify === 'deficit_zero') {
        // Check that beta1 at this edge is zero
        if (this.b1 > 0) {
          this.diagnostics.push({
            line: 1,
            column: 1,
            code: 'ERR_DEFICIT_NONZERO',
            message: `verify: 'deficit_zero' failed: beta1=${this.b1} at ${
              edge.type
            } edge from [${edge.sourceIds.join(', ')}].`,
            severity: 'error',
          });
        }
      } else if (verify === 'first_law') {
        // Already checked globally, make it an error here
        let forkEntropy = 0;
        for (const e of this.ast.edges) {
      if (e.type === 'FORK')
            forkEntropy += Math.log2(Math.max(1, e.targetIds.length));
        }
        const erasure = this.computeLandauerHeat();
        if (forkEntropy > 0 && forkEntropy < erasure - 1e-9) {
          this.diagnostics.push({
            line: 1,
            column: 1,
            code: 'THERMO_FIRST_LAW_VIOLATED',
            message: `verify: 'first_law' failed: fork entropy (${forkEntropy.toFixed(
              3
            )}) < erasure (${erasure.toFixed(3)}).`,
            severity: 'error',
          });
        }
      }
      // convergence and conservation are handled by existing checks
    }
  }

  // ============================================================================
  // Phase 7: Aleph and Entropy Reversal
  // ============================================================================

  private checkAlephCompleteness(): void {
    for (const node of this.ast.nodes.values()) {
      if (!node.labels.some((l) => l === 'Aleph')) continue;

      // Verify Aleph node receives all upstream void data
      const incomingEdges = this.ast.edges.filter((e) =>
        e.targetIds.some((t) => t.trim() === node.id)
      );

      if (incomingEdges.length === 0) {
        this.diagnostics.push({
          line: 1,
          column: 1,
          code: 'ALEPH_INCOMPLETE',
          message: `Aleph node '${node.id}' has no incoming edges. The Aleph must receive all upstream void data to compute the sufficient statistic.`,
          severity: 'error',
        });
        continue;
      }

      // Check that all fork paths eventually reach the Aleph
      const forkTargets = new Set<string>();
      for (const edge of this.ast.edges) {
        if (edge.type === 'FORK') {
          edge.targetIds.forEach((t) => forkTargets.add(t.trim()));
        }
      }

      if (forkTargets.size > 0) {
        const alephSources = new Set<string>();
        const collectUpstream = (
          nodeId: string,
          visited: Set<string>
        ): void => {
          if (visited.has(nodeId)) return;
          visited.add(nodeId);
          alephSources.add(nodeId);
          const incoming = this.ast.edges.filter((e) =>
            e.targetIds.some((t) => t.trim() === nodeId)
          );
          for (const e of incoming) {
            for (const s of e.sourceIds) {
              collectUpstream(s.trim(), visited);
            }
          }
        };
        collectUpstream(node.id, new Set());

        const unreachedForks = [...forkTargets].filter(
          (t) => !alephSources.has(t)
        );
        if (unreachedForks.length > 0) {
          this.diagnostics.push({
            line: 1,
            column: 1,
            code: 'ALEPH_INCOMPLETE',
            message: `Aleph node '${
              node.id
            }' does not receive void data from fork branches: ${unreachedForks.join(
              ', '
            )}. The sufficient statistic requires all upstream void data.`,
            severity: 'warning',
          });
        }
      }
    }
  }

  private checkEntropyReversal(): void {
    // Detect METACOG loops where entropy might reverse
    const metacogEdges = this.ast.edges.filter((e) => e.type === 'METACOG');
    for (const edge of metacogEdges) {
      // A METACOG edge that targets a node which also sources a FORK
      // could cause entropy reversal (information creation in a feedback loop)
      for (const targetId of edge.targetIds) {
        const downstream = this.ast.edges.filter(
          (e) =>
            e.sourceIds.some((s) => s.trim() === targetId.trim()) &&
            e.type === 'FORK'
        );
        if (downstream.length > 0) {
          this.diagnostics.push({
            line: 1,
            column: 1,
            code: 'ENTROPY_REVERSAL_DETECTED',
            message: `METACOG loop at '${targetId.trim()}' feeds into FORK, creating potential entropy reversal. The feedback loop may create information faster than the void boundary can accumulate rejections.`,
            severity: 'info',
          });
        }
      }
    }
  }

  private hasUpstreamVoidEvidence(nodeId: string, seen: Set<string>): boolean {
    if (seen.has(nodeId)) {
      return false;
    }
    seen.add(nodeId);

    const node = this.ast.nodes.get(nodeId);
    if (!node) {
      return false;
    }
    if (
      node.properties.void_boundary !== undefined ||
      node.properties.void_density !== undefined
    ) {
      return true;
    }

    const incomingEdges = this.ast.edges.filter((edge) =>
      edge.targetIds.map((targetId) => targetId.trim()).includes(nodeId)
    );

    for (const edge of incomingEdges) {
      if (
        edge.type === 'FORK' ||
        edge.type === 'RACE' ||
        edge.type === 'SUPERPOSE' ||
        edge.type === 'VENT' ||
        edge.type === 'SLIVER'
      ) {
        return true;
      }

      if (
        edge.type === 'PROCESS' ||
        edge.type === 'LAMINAR' ||
        edge.type === 'OBSERVE' ||
        edge.type === 'MEASURE'
      ) {
        for (const sourceId of edge.sourceIds) {
          if (this.hasUpstreamVoidEvidence(sourceId.trim(), seen)) {
            return true;
          }
        }
      }
    }

    return false;
  }
}
