// @a0n/.ignored_gnosis/src/betty/data-hologram.ts
/**
 * Topological Data Hologram Compressor
 * 
 * Compresses the structural schema of CSV, JSON, and XML into 
 * the 28-byte Grassmannian Amplituhedron tensor, while extracting
 * the raw string/primitive values into a topological dictionary payload.
 */

import type { AmplituhedronGeometry } from '../betti/amplituhedron-volume';

export interface HologramPayload {
  geometry: AmplituhedronGeometry;
  ventDictionary: string[];
}

/**
 * Spatial hash function for schema signatures
 */
function hashSchema(schemaStr: string): number {
  let hash = 0;
  for (let i = 0; i < schemaStr.length; i++) {
    const char = schemaStr.charCodeAt(i);
    hash = ((hash << 5) - hash) + char;
    hash = hash & hash; // Convert to 32bit integer
  }
  return Math.abs(hash);
}

// ==========================================
// CSV / TSV HOLOGRAPHY
// ==========================================

/**
 * Handles the gnosis encode Csv Hologram workflow.
 */
export function encodeCsvHologram(csvText: string, delimiter: string = ','): HologramPayload {
  const lines = csvText.split('\n').map(l => l.trim()).filter(l => l.length > 0);
  if (lines.length === 0) throw new Error("Empty CSV");

  const headers = lines[0]!.split(delimiter);
  const C = headers.length; // dimensionality (k)
  const R = lines.length - 1; // data rows
  const V = R * C; // total static volume (n * k)

  const originVertex = hashSchema(headers.join(delimiter));
  
  let flattenedFaces = 0; // defects (empty cells)
  const ventDictionary: string[] = []; // extracted entropy

  for (let r = 1; r <= R; r++) {
    const cells = lines[r]!.split(delimiter);
    for (let c = 0; c < C; c++) {
      const cell = cells[c] || '';
      if (cell === '') {
        flattenedFaces++; // Defect in the matrix
      }
      ventDictionary.push(cell);
    }
  }

  return {
    geometry: {
      dimensionality: C,
      staticVolume: V,
      originVertex: originVertex,
      flattenedFaces: flattenedFaces,
    },
    ventDictionary
  };
}

/**
 * Handles the gnosis inflate Csv Hologram workflow.
 */
export function inflateCsvHologram(
  hologram: HologramPayload, 
  originalHeaders: string[], 
  delimiter: string = ','
): string {
  const { geometry, ventDictionary } = hologram;
  
  const C = geometry.dimensionality;
  const V = geometry.staticVolume;
  const R = V / C; // mathematically reconstruct row count
  
  const lines: string[] = [];
  lines.push(originalHeaders.join(delimiter));

  let dictIndex = 0;
  for (let r = 0; r < R; r++) {
    const rowCells: string[] = [];
    for (let c = 0; c < C; c++) {
      rowCells.push(ventDictionary[dictIndex]!);
      dictIndex++;
    }
    lines.push(rowCells.join(delimiter));
  }

  return lines.join('\n');
}

// ==========================================
// JSON HOLOGRAPHY (Homogeneous Arrays)
// ==========================================

/**
 * Handles the gnosis encode Json Array Hologram workflow.
 */
export function encodeJsonArrayHologram(jsonArray: Record<string, any>[]): HologramPayload {
  if (jsonArray.length === 0) throw new Error("Empty JSON Array");
  
  // Assuming homogeneous schema based on first element
  const keys = Object.keys(jsonArray[0]!).sort();
  const k = keys.length; // dimensionality
  const n = jsonArray.length; // constraints
  const V = k * n; // static volume

  const originVertex = hashSchema(keys.join(':'));
  
  let flattenedFaces = 0;
  const ventDictionary: string[] = [];

  for (const obj of jsonArray) {
    for (const key of keys) {
      const val = obj[key];
      if (val === null || val === undefined) {
        flattenedFaces++;
        ventDictionary.push(""); // empty placeholder
      } else {
        ventDictionary.push(String(val));
      }
    }
  }

  return {
    geometry: {
      dimensionality: k,
      staticVolume: V,
      originVertex: originVertex,
      flattenedFaces: flattenedFaces,
    },
    ventDictionary
  };
}

/**
 * Handles the gnosis inflate Json Array Hologram workflow.
 */
export function inflateJsonArrayHologram(
  hologram: HologramPayload,
  schemaKeys: string[]
): Record<string, any>[] {
  const { geometry, ventDictionary } = hologram;
  
  const k = geometry.dimensionality;
  const V = geometry.staticVolume;
  const n = V / k;
  
  // The schema must match the dimensionality
  if (schemaKeys.length !== k) {
    throw new Error(`Schema mismatch: expected ${k} keys, got ${schemaKeys.length}`);
  }

  const sortedKeys = [...schemaKeys].sort();
  const result: Record<string, any>[] = [];

  let dictIndex = 0;
  for (let i = 0; i < n; i++) {
    const obj: Record<string, any> = {};
    for (const key of sortedKeys) {
      const val = ventDictionary[dictIndex]!;
      // Simple parse attempt for strings vs numbers/booleans
      if (val === "") {
        obj[key] = null;
      } else if (val === "true") {
        obj[key] = true;
      } else if (val === "false") {
        obj[key] = false;
      } else if (!isNaN(Number(val)) && val.trim() !== '') {
        obj[key] = Number(val);
      } else {
        obj[key] = val;
      }
      dictIndex++;
    }
    result.push(obj);
  }

  return result;
}

// ==========================================
// MARKDOWN HOLOGRAPHY (Block Topology)
// ==========================================

// Simple deterministic block typing
function getMarkdownBlockType(block: string): string {
  if (block.startsWith('#')) return 'heading';
  if (block.startsWith('- ') || block.startsWith('* ')) return 'list';
  if (block.startsWith('>')) return 'blockquote';
  if (block.startsWith('```')) return 'code';
  return 'paragraph';
}

export function encodeMarkdownHologram(mdText: string): HologramPayload {
  const blocks = mdText.split('\n\n').map(b => b.trim());
  if (blocks.length === 0) throw new Error("Empty Markdown");

  // We map the document to a 5-dimensional block space
  const blockTypes = ['heading', 'list', 'blockquote', 'code', 'paragraph'];
  const k = blockTypes.length; // dimensionality
  const n = blocks.length; // total constraints/blocks
  const V = k * n; // static volume

  // The origin vertex is the structural sequence of the document
  const blockSequence = blocks.map(b => getMarkdownBlockType(b)).join(':');
  const originVertex = hashSchema(blockSequence);

  let flattenedFaces = 0;
  const ventDictionary: string[] = [];

  for (const block of blocks) {
    if (block === '') {
      flattenedFaces++; // empty block defect
      ventDictionary.push('');
    } else {
      ventDictionary.push(block);
    }
  }

  return {
    geometry: {
      dimensionality: k,
      staticVolume: V,
      originVertex: originVertex,
      flattenedFaces: flattenedFaces,
    },
    ventDictionary
  };
}

export function inflateMarkdownHologram(
  hologram: HologramPayload,
  structuralSequence: string
): string {
  const { geometry, ventDictionary } = hologram;
  
  const k = geometry.dimensionality;
  const V = geometry.staticVolume;
  const n = V / k;
  
  const expectedSequence = structuralSequence.split(':');
  if (expectedSequence.length !== n) {
    throw new Error(`Sequence mismatch: expected ${n} blocks, got ${expectedSequence.length}`);
  }

  const blocks: string[] = [];
  for (let i = 0; i < n; i++) {
    blocks.push(ventDictionary[i]!);
  }

  return blocks.join('\n\n');
}
