// @a0n/.ignored_gnosis/src/betty/fano-plane.ts
/**
 * The Fano Plane Substrate
 * 
 * Defines the multiplication rules for the 7 imaginary basis vectors of the Octonions (O).
 * The Octonions are the only normed division algebra in 8 dimensions.
 * They are fundamentally non-associative: (e_i * e_j) * e_k !== e_i * (e_j * e_k).
 */

export type OctonionBasis = 'e1' | 'e2' | 'e3' | 'e4' | 'e5' | 'e6' | 'e7';

export interface OctonionState {
  basis: OctonionBasis;
  sign: 1 | -1;
}

// The 7 projective lines of the Fano Plane.
// Each line forms a Quaternionic sub-algebra triad.
// Directed cyclic triads:
const FANO_TRIADS = [
  ['e1', 'e2', 'e4'],
  ['e2', 'e3', 'e5'],
  ['e3', 'e4', 'e6'],
  ['e4', 'e5', 'e7'],
  ['e5', 'e6', 'e1'],
  ['e6', 'e7', 'e2'],
  ['e7', 'e1', 'e3'],
] as const;

/**
 * Multiplies two Octonionic basis vectors using the Fano Plane geometry.
 */
export function fanoMultiply(a: OctonionState, b: OctonionState): OctonionState {
  const signMultiplier = (a.sign * b.sign) as 1 | -1;

  if (a.basis === b.basis) {
    // e_i * e_i = -1 (Real number). 
    // For pure execution routing, we map real collapse to an anomalous state or identity,
    // but typically we want to avoid total collapse. Let's return a "Real" collapse error 
    // or map it back into the imaginary plane for continuous routing.
    throw new Error("Topological Collapse: Multiplication of identical basis vectors yields a Real number (-1).");
  }

  for (const triad of FANO_TRIADS) {
    const [x, y, z] = triad;
    
    if (a.basis === x && b.basis === y) return { basis: z!, sign: signMultiplier };
    if (a.basis === y && b.basis === z) return { basis: x!, sign: signMultiplier };
    if (a.basis === z && b.basis === x) return { basis: y!, sign: signMultiplier };

    // Anti-commutative reversing
    if (b.basis === x && a.basis === y) return { basis: z!, sign: (signMultiplier * -1) as 1 | -1 };
    if (b.basis === y && a.basis === z) return { basis: x!, sign: (signMultiplier * -1) as 1 | -1 };
    if (b.basis === z && a.basis === x) return { basis: y!, sign: (signMultiplier * -1) as 1 | -1 };
  }

  throw new Error(`Invalid Fano Plane pairing: ${a.basis} x ${b.basis}`);
}

/**
 * Creates a raw basis vector state.
 */
export function e(basis: OctonionBasis, sign: 1 | -1 = 1): OctonionState {
  return { basis, sign };
}
