// @a0n/.ignored_gnosis/src/betty/hologram-query.ts
/**
 * Schubert Query Compiler
 * 
 * Intercepts Gnosis (.gg) topologies and maps them directly to 
 * Topological Raycaster projections against a Hologram.
 */

import type { GraphAST } from './compiler';
import type { HologramPayload } from './data-hologram';
import { HologramNavigator } from './hologram-navigator';
import { OctonionicRouter, type OpGraph } from './octonionic-router';

export interface SchubertQueryResult {
  type: 'AFFINE_SCALAR' | 'SCHUBERT_SUBSET';
  scalar?: string;
  subset?: HologramPayload;
}

/**
 * Executes a pre-parsed .gg Query AST against a source Hologram.
 */
export function executeHologramQuery(
  queryAst: GraphAST, 
  sourceHologram: HologramPayload
): SchubertQueryResult[] {
  const navigator = new HologramNavigator(sourceHologram);
  const results: SchubertQueryResult[] = [];

  for (const edge of queryAst.edges) {
    // 1. Layer 2: [:OBSERVE] -> Affine Projection
    if (edge.type === 'OBSERVE') {
      const row = Number(edge.properties['row']);
      const col = Number(edge.properties['col']);
      
      if (!isNaN(row) && !isNaN(col)) {
        const val = navigator.projectAffine(row, col);
        results.push({ type: 'AFFINE_SCALAR', scalar: val });
      }
    }

    // 2. Layer 2: [:SLIVER] -> Schubert Slicing
    if (edge.type === 'SLIVER') {
      const targetCol = Number(edge.properties['col']);
      const operator = String(edge.properties['operator']);
      const targetValue = edge.properties['value'];

      if (!isNaN(targetCol) && operator && targetValue !== undefined) {
        const subset = navigator.intersectSchubert((val, colIndex) => {
          if (colIndex !== targetCol) return false;
          
          // Apply mathematical slicing operator
          if (operator === '>') return Number(val) > Number(targetValue);
          if (operator === '<') return Number(val) < Number(targetValue);
          if (operator === '==') return val == targetValue;
          if (operator === '!=') return val != targetValue;
          return false;
        }, { col: targetCol, op: operator, val: targetValue });

        results.push({ type: 'SCHUBERT_SUBSET', subset });
      }
    }
  }

  return results;
}

/**
 * Executes a non-associative query graph directly via the Octonionic Compiler.
 * Eliminates sequential processing entirely.
 */
export function executeOctonionicQuery(
  graph: OpGraph, 
  hologram: HologramPayload, 
  targetCol: number, 
  targetValue: string
): HologramPayload {
  const router = new OctonionicRouter(hologram);
  return router.routeExecution(graph, targetCol, targetValue);
}
