// @a0n/.ignored_gnosis/src/betty/legalese.ts
export type LegaleseClauseKind =
  | 'definition'
  | 'user_obligation'
  | 'user_prohibition'
  | 'user_right'
  | 'provider_right'
  | 'provider_obligation'
  | 'limitation_of_liability'
  | 'dispute_resolution'
  | 'termination'
  | 'payment'
  | 'privacy_data'
  | 'content_license'
  | 'general';

export type LegaleseRiskKind =
  | 'unilateral_change'
  | 'termination_or_suspension'
  | 'liability_limit'
  | 'mandatory_arbitration'
  | 'jurisdiction_governing_law'
  | 'content_license'
  | 'payment_or_auto_renewal'
  | 'data_or_privacy'
  | 'third_party_dependency'
  | 'unclear_language';

export type LegaleseActor =
  | 'user'
  | 'provider'
  | 'third_party'
  | 'regulator'
  | 'both'
  | 'unknown';

export type LegaleseClarity = 'plain' | 'moderate' | 'hard' | 'opaque';

export type LegaleseSpanKind =
  | 'clause'
  | 'heading'
  | 'definition'
  | 'cross_reference'
  | 'condition'
  | 'exception'
  | 'frame'
  | 'hole';

export type LegaleseDeonticOperator =
  | 'may'
  | 'must'
  | 'must_not'
  | 'will'
  | 'can'
  | 'waives'
  | 'disclaims'
  | 'limits'
  | 'consents'
  | 'grants'
  | 'terminates'
  | 'changes'
  | 'collects'
  | 'shares'
  | 'pays'
  | 'refunds'
  | 'renews'
  | 'governs'
  | 'unknown';

export type LegaleseLegalAction =
  | 'change'
  | 'terminate'
  | 'suspend'
  | 'delete'
  | 'disclaim'
  | 'limit'
  | 'arbitrate'
  | 'waive'
  | 'grant_license'
  | 'collect'
  | 'share'
  | 'process'
  | 'pay'
  | 'charge'
  | 'refund'
  | 'renew'
  | 'govern'
  | 'choose_forum'
  | 'incorporate_document'
  | 'interpret'
  | 'unknown';

export type LegaleseLegalObject =
  | 'terms'
  | 'fees'
  | 'service'
  | 'account'
  | 'user_content'
  | 'personal_data'
  | 'liability'
  | 'warranty'
  | 'damages'
  | 'dispute'
  | 'jurisdiction'
  | 'law'
  | 'document'
  | 'feedback'
  | 'unknown';

export type LegaleseFrameHoleKind =
  | 'actor'
  | 'action'
  | 'object'
  | 'scope'
  | 'condition'
  | 'definition'
  | 'cross_reference'
  | 'exception';

export interface LegaleseSpan {
  kind: LegaleseSpanKind;
  startOffset: number;
  endOffset: number;
  text: string;
}

export interface LegaleseSection {
  id: string;
  index: number;
  path: string[];
  heading: string;
  startOffset: number;
  endOffset: number;
}

export interface LegaleseDefinition {
  id: string;
  term: string;
  meaning: string;
  span: LegaleseSpan;
}

export interface LegaleseCrossReference {
  id: string;
  rawText: string;
  target: string;
  span: LegaleseSpan;
  resolved: boolean;
  resolvedSectionId?: string;
  resolvedSectionPath?: string[];
}

export interface LegaleseFrameHole {
  id: string;
  kind: LegaleseFrameHoleKind;
  message: string;
  span: LegaleseSpan;
}

export interface LegaleseClauseFrame {
  id: string;
  riskKind: LegaleseRiskKind;
  actor: LegaleseActor;
  recipient: LegaleseActor;
  operator: LegaleseDeonticOperator;
  action: LegaleseLegalAction;
  object: LegaleseLegalObject;
  conditions: string[];
  exceptions: string[];
  scope: string[];
  evidenceSpans: LegaleseSpan[];
  holes: LegaleseFrameHole[];
  confidence: number;
}

export interface LegaleseLoweringOptions {
  documentId?: string;
  sourcePath?: string;
  serviceName?: string;
  maxClauses?: number;
  minClauseChars?: number;
  maxClauseChars?: number;
  excerptChars?: number;
}

export interface LegaleseClauseSignal {
  kind: LegaleseRiskKind | LegaleseClauseKind;
  confidence: number;
  pattern: string;
}

export interface LegaleseClauseLowering {
  id: string;
  index: number;
  text: string;
  excerpt: string;
  kind: LegaleseClauseKind;
  actor: LegaleseActor;
  risks: LegaleseRiskKind[];
  signals: LegaleseClauseSignal[];
  wordCount: number;
  sentenceCount: number;
  averageWordsPerSentence: number;
  legalTermCount: number;
  clarity: LegaleseClarity;
  clarityScore: number;
  sectionPath: string[];
  heading: string;
  spans: LegaleseSpan[];
  definitions: LegaleseDefinition[];
  crossReferences: LegaleseCrossReference[];
  frames: LegaleseClauseFrame[];
  holes: LegaleseFrameHole[];
}

export interface LegaleseLoweringSummary {
  documentId: string;
  serviceName: string;
  clauseCount: number;
  riskCounts: Record<LegaleseRiskKind, number>;
  hardClauseCount: number;
  opaqueClauseCount: number;
  averageWordsPerClause: number;
}

export interface LegaleseLoweringResult {
  ggSource: string;
  documentId: string;
  serviceName: string;
  clauses: LegaleseClauseLowering[];
  sections: LegaleseSection[];
  risks: LegaleseRiskKind[];
  summary: LegaleseLoweringSummary;
}

export interface LegaleseClauseSegment {
  text: string;
  startOffset: number;
  endOffset: number;
  sectionPath: string[];
  heading: string;
  blockIndex: number;
  sentenceIndex: number;
}

export interface LegaleseSegmentationResult {
  normalizedText: string;
  sections: LegaleseSection[];
  clauses: LegaleseClauseSegment[];
}

interface ClauseKindRule {
  kind: LegaleseClauseKind;
  actor: LegaleseActor;
  confidence: number;
  pattern: RegExp;
}

interface RiskRule {
  kind: LegaleseRiskKind;
  confidence: number;
  pattern: RegExp;
}

interface SourceLine {
  text: string;
  startOffset: number;
  endOffset: number;
}

interface SourceSpan {
  text: string;
  startOffset: number;
  endOffset: number;
}

const DEFAULT_MIN_CLAUSE_CHARS = 32;
const DEFAULT_MAX_CLAUSE_CHARS = 1200;
const DEFAULT_EXCERPT_CHARS = 180;

const LEGAL_TERMS = [
  'agreement',
  'arbitration',
  'assign',
  'breach',
  'claim',
  'consent',
  'damages',
  'disclaimer',
  'dispute',
  'indemnify',
  'jurisdiction',
  'liable',
  'liability',
  'license',
  'notwithstanding',
  'obligation',
  'provision',
  'pursuant',
  'remedy',
  'terminate',
  'terms',
  'warranty',
];

const CLAUSE_KIND_RULES: ClauseKindRule[] = [
  {
    kind: 'user_prohibition',
    actor: 'user',
    confidence: 0.88,
    pattern: /\b(you|user|users)\s+(?:may|must|shall|can)\s+not\b|\bprohibited\b|\bnot\s+permitted\b/i,
  },
  {
    kind: 'user_obligation',
    actor: 'user',
    confidence: 0.82,
    pattern: /\b(you|user|users)\s+(?:must|shall|agree|are\s+responsible|required|need)\b/i,
  },
  {
    kind: 'provider_right',
    actor: 'provider',
    confidence: 0.84,
    pattern: /\b(we|company|service|platform|provider|fintecom)\s+(?:may|can|reserve|retains?|have\s+the\s+right)\b/i,
  },
  {
    kind: 'provider_obligation',
    actor: 'provider',
    confidence: 0.72,
    pattern: /\b(we|company|service|platform|provider|fintecom)\s+(?:will|shall|must|are\s+required)\b/i,
  },
  {
    kind: 'limitation_of_liability',
    actor: 'provider',
    confidence: 0.9,
    pattern: /\b(no|not)\s+liab(?:le|ility)\b|\blimitation\s+of\s+liability\b|\bwithout\s+warrant(?:y|ies)\b|\bas\s+is\b/i,
  },
  {
    kind: 'dispute_resolution',
    actor: 'both',
    confidence: 0.86,
    pattern: /\barbitration\b|\bclass\s+action\b|\bgoverning\s+law\b|\bjurisdiction\b|\bvenue\b/i,
  },
  {
    kind: 'termination',
    actor: 'provider',
    confidence: 0.78,
    pattern: /\bterminate\b|\bsuspend\b|\bdisable\b|\bdelete\s+(?:your\s+)?account\b|\brefuse\s+service\b/i,
  },
  {
    kind: 'payment',
    actor: 'user',
    confidence: 0.76,
    pattern: /\bsubscription\b|\bauto[-\s]?renew\b|\bbilling\b|\brefund\b|\bfee\b|\bcharge\b|\bpaid\b/i,
  },
  {
    kind: 'privacy_data',
    actor: 'both',
    confidence: 0.74,
    pattern: /\bpersonal\s+data\b|\bprivacy\s+policy\b|\bcollect\b|\bprocess(?:ing)?\b|\bshare\s+(?:your\s+)?(?:data|information)\b/i,
  },
  {
    kind: 'content_license',
    actor: 'user',
    confidence: 0.82,
    pattern: /\buser\s+content\b|\broyalty[-\s]?free\b|\bworldwide\b|\bsublicens(?:e|able)\b|\blicense\s+(?:to\s+)?(?:use|host|copy|display)\b/i,
  },
  {
    kind: 'definition',
    actor: 'unknown',
    confidence: 0.64,
    pattern: /\bhereinafter\b|\bmeans\b|\breferred\s+to\s+as\b|\bdefined\s+as\b/i,
  },
  {
    kind: 'user_right',
    actor: 'user',
    confidence: 0.58,
    pattern: /\b(you|user|users)\s+(?:may|can|are\s+entitled)\b/i,
  },
];

const RISK_RULES: RiskRule[] = [
  {
    kind: 'unilateral_change',
    confidence: 0.88,
    pattern: /\b(?:we|company|service|platform|provider)\b.{0,80}\b(?:change|modify|update|revise)\b.{0,80}\b(?:terms|agreement|service|fees|account|content)\b|\bterms\b.{0,60}\b(?:change|modify|update|revise)\b|\b(?:change|modify|update|revise)\b.{0,80}\b(?:terms|agreement|service|fees)\b/i,
  },
  {
    kind: 'termination_or_suspension',
    confidence: 0.82,
    pattern: /\bterminate\b|\bsuspend\b|\bdisable\b|\bdelete\s+(?:your\s+)?account\b|\brefuse\s+service\b/i,
  },
  {
    kind: 'liability_limit',
    confidence: 0.9,
    pattern: /\b(no|not)\s+liab(?:le|ility)\b|\blimitation\s+of\s+liability\b|\bdisclaim(?:er)?\b|\bwithout\s+warrant(?:y|ies)\b|\bas\s+is\b/i,
  },
  {
    kind: 'mandatory_arbitration',
    confidence: 0.86,
    pattern: /\barbitration\b|\bclass\s+action\b|\bjury\s+trial\b|\bcollective\s+action\b/i,
  },
  {
    kind: 'jurisdiction_governing_law',
    confidence: 0.72,
    pattern: /\bgoverning\s+law\b|\bjurisdiction\b|\bvenue\b|\blaws\s+of\b|\bcourts?\s+of\b/i,
  },
  {
    kind: 'content_license',
    confidence: 0.82,
    pattern: /\buser\s+content\b|\broyalty[-\s]?free\b|\bworldwide\b|\bsublicens(?:e|able)\b|\blicense\s+(?:to\s+)?(?:use|host|copy|display)\b/i,
  },
  {
    kind: 'payment_or_auto_renewal',
    confidence: 0.76,
    pattern: /\bsubscription\b|\bauto[-\s]?renew\b|\bbilling\b|\brefund\b|\bfee\b|\bcharge\b|\bpaid\b/i,
  },
  {
    kind: 'data_or_privacy',
    confidence: 0.74,
    pattern: /\bpersonal\s+data\b|\bpersonal\s+information\b|\bprivacy\s+policy\b|\bcollect\b|\bprocess(?:ing)?\b|\bshare\s+(?:your\s+)?(?:data|information)\b/i,
  },
  {
    kind: 'third_party_dependency',
    confidence: 0.66,
    pattern: /\bthird[-\s]?party\b|\baffiliate\b|\bpartner\b|\bexternal\s+link\b|\bprivacy\s+policy\b|\buser\s+guidelines\b/i,
  },
];

const ALL_RISK_KINDS: LegaleseRiskKind[] = [
  'unilateral_change',
  'termination_or_suspension',
  'liability_limit',
  'mandatory_arbitration',
  'jurisdiction_governing_law',
  'content_license',
  'payment_or_auto_renewal',
  'data_or_privacy',
  'third_party_dependency',
  'unclear_language',
];

/**
 * Handles the gnosis lower Legalese To Gg workflow.
 */
export function lowerLegaleseToGg(
  input: string,
  options: LegaleseLoweringOptions = {}
): LegaleseLoweringResult {
  const serviceName = options.serviceName ?? inferServiceName(input);
  const documentId = toNodeId(options.documentId ?? serviceName, 'tos');
  const segmented = segmentLegaleseDetailed(input, options);
  const excerptChars = options.excerptChars ?? DEFAULT_EXCERPT_CHARS;

  const clauses = segmented.clauses.map((segment, index) =>
    lowerClause(segment, index, documentId, excerptChars, segmented.sections)
  );
  const risks = uniqueRiskKinds(clauses.flatMap((clause) => clause.risks));
  const summary = buildSummary(documentId, serviceName, clauses);
  const ggSource = emitLegaleseGg(
    documentId,
    serviceName,
    options,
    segmented.sections,
    clauses,
    risks,
    summary
  );

  return {
    ggSource,
    documentId,
    serviceName,
    clauses,
    sections: segmented.sections,
    risks,
    summary,
  };
}

export const compileLegaleseToGg = lowerLegaleseToGg;

/**
 * Handles the gnosis segment Legalese workflow.
 */
export function segmentLegalese(
  input: string,
  options: Pick<
    LegaleseLoweringOptions,
    'maxClauses' | 'minClauseChars' | 'maxClauseChars'
  > = {}
): string[] {
  return segmentLegaleseDetailed(input, options).clauses.map((clause) => clause.text);
}

/**
 * Handles the gnosis segment Legalese Detailed workflow.
 */
export function segmentLegaleseDetailed(
  input: string,
  options: Pick<
    LegaleseLoweringOptions,
    'maxClauses' | 'minClauseChars' | 'maxClauseChars'
  > = {}
): LegaleseSegmentationResult {
  const minClauseChars = options.minClauseChars ?? DEFAULT_MIN_CLAUSE_CHARS;
  const maxClauseChars = options.maxClauseChars ?? DEFAULT_MAX_CLAUSE_CHARS;
  const normalizedText = normalizeLegaleseSource(input);

  if (!normalizedText) {
    return { normalizedText, sections: [], clauses: [] };
  }

  const sections: LegaleseSection[] = [];
  const clauses: LegaleseClauseSegment[] = [];
  const lines = splitLinesWithOffsets(normalizedText);
  let activePath: string[] = [];
  let activeHeading = '';
  let block: SourceLine[] = [];
  let blockIndex = 0;

  const flushBlock = () => {
    if (block.length === 0) {
      return;
    }

    const blockSpan = spanFromLines(block);
    const blockText = blockSpan.text.trim();
    if (blockText.length === 0) {
      block = [];
      return;
    }

    const heading = parseHeadingBlock(blockText, block);
    if (heading) {
      activePath = heading.path;
      activeHeading = heading.heading;
      sections.push({
        id: `section_${sections.length + 1}`,
        index: sections.length,
        path: activePath,
        heading: activeHeading,
        startOffset: blockSpan.startOffset,
        endOffset: blockSpan.endOffset,
      });
      block = [];
      blockIndex++;
      return;
    }

    const candidateSpans = splitBlockIntoCandidateSpans(blockSpan, block);
    let sentenceIndex = 0;
    for (const candidate of candidateSpans) {
      for (const sentence of splitSentencesWithOffsets(candidate)) {
        for (const clauseSpan of splitOversizedClauseWithOffsets(sentence, maxClauseChars)) {
          const trimmed = trimSourceSpan(clauseSpan);
          const cleaned = normalizeClauseText(trimmed.text);
          if (cleaned.length >= minClauseChars) {
            clauses.push({
              text: cleaned,
              startOffset: trimmed.startOffset,
              endOffset: trimmed.endOffset,
              sectionPath: activePath,
              heading: activeHeading,
              blockIndex,
              sentenceIndex,
            });
            sentenceIndex++;
            if (options.maxClauses && clauses.length >= options.maxClauses) {
              block = [];
              return;
            }
          }
        }
      }
    }

    block = [];
    blockIndex++;
  };

  for (const line of lines) {
    if (line.text.trim().length === 0) {
      flushBlock();
      if (options.maxClauses && clauses.length >= options.maxClauses) {
        break;
      }
      continue;
    }
    block.push(line);
  }
  if (!options.maxClauses || clauses.length < options.maxClauses) {
    flushBlock();
  }

  return { normalizedText, sections, clauses };
}

function lowerClause(
  segment: LegaleseClauseSegment,
  index: number,
  documentId: string,
  excerptChars: number,
  sections: LegaleseSection[]
): LegaleseClauseLowering {
  const text = segment.text;
  const clauseId = `${documentId}_clause_${String(index + 1).padStart(4, '0')}`;
  const classification = classifyClause(text);
  const words = tokenizeWords(text);
  const sentenceCount = Math.max(1, splitSentences(text).length);
  const averageWordsPerSentence = words.length / sentenceCount;
  const legalTermCount = countLegalTerms(words);
  const clarityScore = scoreClarity(text, words, averageWordsPerSentence, legalTermCount);
  const clarity = classifyClarity(clarityScore);
  const definitions = extractDefinitions(clauseId, text, segment.startOffset);
  const crossReferences = extractCrossReferences(clauseId, text, segment.startOffset, sections);
  const frameRisks = uniqueRiskKinds(classification.risks);
  const frames = extractClauseFrames(
    clauseId,
    text,
    segment.startOffset,
    classification.actor,
    frameRisks
  );
  const risks = uniqueRiskKinds([...classification.risks, ...frames.map((frame) => frame.riskKind)]);
  const signals = [...classification.signals];
  const crossReferenceHoles = crossReferences
    .filter((reference) => !reference.resolved && shouldRequireResolvedReference(reference.rawText))
    .map((reference, holeIndex) =>
      buildFrameHole(
        `${clauseId}_xref_hole_${holeIndex + 1}`,
        'cross_reference',
        `Unresolved legal cross-reference: ${reference.rawText}`,
        reference.span
      )
    );
  const holes = [...frames.flatMap((frame) => frame.holes), ...crossReferenceHoles];

  if ((clarity === 'hard' || clarity === 'opaque') && !risks.includes('unclear_language')) {
    risks.push('unclear_language');
    signals.push({
      kind: 'unclear_language',
      confidence: clarity === 'opaque' ? 0.88 : 0.72,
      pattern: 'readability-threshold',
    });
  }

  return {
    id: clauseId,
    index,
    text,
    excerpt: sanitizePropertyText(text, excerptChars),
    kind: classification.kind,
    actor: classification.actor,
    risks: uniqueRiskKinds(risks),
    signals,
    wordCount: words.length,
    sentenceCount,
    averageWordsPerSentence,
    legalTermCount,
    clarity,
    clarityScore,
    sectionPath: segment.sectionPath,
    heading: segment.heading,
    spans: [
      {
        kind: 'clause',
        startOffset: segment.startOffset,
        endOffset: segment.endOffset,
        text,
      },
    ],
    definitions,
    crossReferences,
    frames,
    holes,
  };
}

function classifyClause(text: string): {
  kind: LegaleseClauseKind;
  actor: LegaleseActor;
  risks: LegaleseRiskKind[];
  signals: LegaleseClauseSignal[];
} {
  const kindMatch = CLAUSE_KIND_RULES.find((rule) => rule.pattern.test(text));
  const riskMatches = RISK_RULES.filter((rule) => rule.pattern.test(text));
  const signals: LegaleseClauseSignal[] = [];

  if (kindMatch) {
    signals.push({
      kind: kindMatch.kind,
      confidence: kindMatch.confidence,
      pattern: kindMatch.pattern.source,
    });
  }

  for (const risk of riskMatches) {
    signals.push({
      kind: risk.kind,
      confidence: risk.confidence,
      pattern: risk.pattern.source,
    });
  }

  return {
    kind: kindMatch?.kind ?? 'general',
    actor: kindMatch?.actor ?? inferActor(text),
    risks: uniqueRiskKinds(riskMatches.map((rule) => rule.kind)),
    signals,
  };
}

function extractClauseFrames(
  clauseId: string,
  text: string,
  startOffset: number,
  fallbackActor: LegaleseActor,
  risks: LegaleseRiskKind[]
): LegaleseClauseFrame[] {
  return risks
    .filter((risk) => risk !== 'unclear_language')
    .map((risk, index) => buildClauseFrame(clauseId, index, text, startOffset, fallbackActor, risk));
}

function buildClauseFrame(
  clauseId: string,
  index: number,
  text: string,
  startOffset: number,
  fallbackActor: LegaleseActor,
  risk: LegaleseRiskKind
): LegaleseClauseFrame {
  const frameId = `${clauseId}_frame_${String(index + 1).padStart(2, '0')}`;
  const actor = inferFrameActor(text, fallbackActor, risk);
  const recipient = inferFrameRecipient(text, risk);
  const operator = inferOperator(text, risk);
  const action = inferAction(text, risk);
  const object = inferObject(text, risk);
  const scope = inferScope(text, risk);
  const conditions = inferConditions(text);
  const exceptions = inferExceptions(text);
  const evidenceSpan = matchEvidenceSpan(text, startOffset, risk);
  const holes = buildFrameHoles(frameId, actor, action, object, scope, evidenceSpan, risk);

  return {
    id: frameId,
    riskKind: risk,
    actor,
    recipient,
    operator,
    action,
    object,
    conditions,
    exceptions,
    scope,
    evidenceSpans: [evidenceSpan],
    holes,
    confidence: riskConfidenceFromText(text, risk),
  };
}

function buildFrameHoles(
  frameId: string,
  actor: LegaleseActor,
  action: LegaleseLegalAction,
  object: LegaleseLegalObject,
  scope: string[],
  evidenceSpan: LegaleseSpan,
  risk: LegaleseRiskKind
): LegaleseFrameHole[] {
  const holes: LegaleseFrameHole[] = [];
  if (actor === 'unknown') {
    holes.push(buildFrameHole(`${frameId}_actor_hole`, 'actor', 'Missing legal actor premise', evidenceSpan));
  }
  if (action === 'unknown') {
    holes.push(buildFrameHole(`${frameId}_action_hole`, 'action', 'Missing legal action premise', evidenceSpan));
  }
  if (object === 'unknown') {
    holes.push(buildFrameHole(`${frameId}_object_hole`, 'object', 'Missing legal object premise', evidenceSpan));
  }
  if (risk === 'content_license' && scope.length === 0) {
    holes.push(buildFrameHole(`${frameId}_scope_hole`, 'scope', 'Missing user-content license scope', evidenceSpan));
  }
  return holes;
}

function buildFrameHole(
  id: string,
  kind: LegaleseFrameHoleKind,
  message: string,
  span: LegaleseSpan
): LegaleseFrameHole {
  return {
    id,
    kind,
    message,
    span: {
      kind: 'hole',
      startOffset: span.startOffset,
      endOffset: span.endOffset,
      text: span.text,
    },
  };
}

function inferFrameActor(
  text: string,
  fallbackActor: LegaleseActor,
  risk: LegaleseRiskKind
): LegaleseActor {
  const lower = text.toLowerCase();
  if (/\b(you|user|users|customer|subscriber)\b/.test(lower)) {
    if (risk === 'content_license' && /\b(grant|license)\b/.test(lower)) {
      return 'user';
    }
    if (/\b(we|us|our|company|provider|platform|service provider)\b/.test(lower)) {
      return 'both';
    }
    if (risk === 'payment_or_auto_renewal' && /\b(pay|fee|charge|billing)\b/.test(lower)) {
      return 'user';
    }
  }
  if (/\b(we|us|our|company|provider|platform|service provider)\b/.test(lower)) {
    return 'provider';
  }
  if (
    risk === 'unilateral_change' &&
    /\b(?:these\s+)?(?:terms|agreement|contract|service)\s+(?:may|can|will)?\s*(?:change|modify|update|revise)\b/.test(lower)
  ) {
    return 'unknown';
  }
  if (/\b[A-Z][A-Za-z0-9&. -]{1,48}\s+(?:may|can|reserves?|will|collects?|charges?|changes?|terminates?|suspends?|disclaims?|limits?)\b/.test(text)) {
    return 'provider';
  }
  if (risk === 'liability_limit' && /\b(as is|without warrant|not liable|no liability|limitation of liability)\b/.test(lower)) {
    return 'provider';
  }
  if (risk === 'mandatory_arbitration' && /\b(dispute|claim|arbitration|class action|jury trial)\b/.test(lower)) {
    return 'both';
  }
  if (risk === 'jurisdiction_governing_law' && /\b(governed by|laws of|courts? of|venue|jurisdiction)\b/.test(lower)) {
    return 'both';
  }
  if (/\b(third[-\s]?party|affiliate|partner)\b/.test(lower)) {
    return 'third_party';
  }
  return fallbackActor;
}

function inferFrameRecipient(text: string, risk: LegaleseRiskKind): LegaleseActor {
  const lower = text.toLowerCase();
  if (risk === 'content_license') {
    if (
      /\bgrant\s+(?:us|our|the\s+company|the\s+provider|[A-Z][A-Za-z0-9&. -]{1,48})\b/.test(text) ||
      /\b(to|for)\s+(?:us|we|our|company|provider|platform|service|[A-Z][A-Za-z0-9&. -]{1,48})\b/.test(text)
    ) {
      return 'provider';
    }
  }
  if (/\bthird[-\s]?party|affiliate|partner\b/.test(lower)) {
    return 'third_party';
  }
  return 'unknown';
}

function inferOperator(text: string, risk: LegaleseRiskKind): LegaleseDeonticOperator {
  const lower = text.toLowerCase();
  if (/\bmay\b|\bcan\b|\breserves?\s+the\s+right\b/.test(lower)) {
    return lower.includes(' can ') ? 'can' : 'may';
  }
  if (/\bmust\s+not\b|\bmay\s+not\b|\bshall\s+not\b|\bwaive\b/.test(lower)) {
    return lower.includes('waive') ? 'waives' : 'must_not';
  }
  if (/\bmust\b|\bshall\b|\bagree\b|\brequired\b/.test(lower)) {
    return 'must';
  }
  if (/\bwill\b/.test(lower)) {
    return 'will';
  }
  switch (risk) {
    case 'unilateral_change':
      return 'changes';
    case 'termination_or_suspension':
      return 'terminates';
    case 'liability_limit':
      return lower.includes('limit') ? 'limits' : 'disclaims';
    case 'content_license':
      return 'grants';
    case 'data_or_privacy':
      return lower.includes('share') ? 'shares' : 'collects';
    case 'payment_or_auto_renewal':
      return lower.includes('renew') ? 'renews' : 'pays';
    case 'jurisdiction_governing_law':
      return 'governs';
    case 'mandatory_arbitration':
      return lower.includes('waive') ? 'waives' : 'must';
    case 'third_party_dependency':
    case 'unclear_language':
      return 'unknown';
  }
}

function inferAction(text: string, risk: LegaleseRiskKind): LegaleseLegalAction {
  const lower = text.toLowerCase();
  if (/\b(change|modify|update|revise)\b/.test(lower)) {
    return 'change';
  }
  if (/\bterminate\b/.test(lower)) {
    return 'terminate';
  }
  if (/\bsuspend\b|\bdisable\b|\brefuse\b/.test(lower)) {
    return 'suspend';
  }
  if (/\bdelete\b|\bremove\b/.test(lower)) {
    return 'delete';
  }
  if (/\bdisclaim\b|\bas\s+is\b|\bwithout\s+warrant/.test(lower)) {
    return 'disclaim';
  }
  if (/\blimit(?:s|ed|ation)?\b|\bnot\s+liable\b|\bno\s+liability\b/.test(lower)) {
    return 'limit';
  }
  if (/\barbitration\b|\barbitrate\b/.test(lower)) {
    return 'arbitrate';
  }
  if (/\bwaive\b|\bclass\s+action\b|\bjury\s+trial\b/.test(lower)) {
    return 'waive';
  }
  if (/\bgrant\b|\blicense\b/.test(lower)) {
    return 'grant_license';
  }
  if (/\bcollect\b/.test(lower)) {
    return 'collect';
  }
  if (/\bshare\b|\bdisclos/.test(lower)) {
    return 'share';
  }
  if (/\bprocess(?:ing)?\b/.test(lower)) {
    return 'process';
  }
  if (/\bpay\b/.test(lower)) {
    return 'pay';
  }
  if (/\bcharge\b|\bfee\b|\bbilling\b/.test(lower)) {
    return 'charge';
  }
  if (/\brefund\b/.test(lower)) {
    return 'refund';
  }
  if (/\brenew\b/.test(lower)) {
    return 'renew';
  }
  if (/\bgoverned\s+by\b|\blaws\s+of\b/.test(lower)) {
    return 'govern';
  }
  if (/\bvenue\b|\bjurisdiction\b|\bcourts?\s+of\b/.test(lower)) {
    return 'choose_forum';
  }
  if (/\bprivacy\s+policy\b|\buser\s+guidelines\b|\bother\s+terms\b/.test(lower)) {
    return 'incorporate_document';
  }
  if (/\binterpret\b/.test(lower)) {
    return 'interpret';
  }
  return risk === 'third_party_dependency' ? 'incorporate_document' : 'unknown';
}

function inferObject(text: string, risk: LegaleseRiskKind): LegaleseLegalObject {
  const lower = text.toLowerCase();
  if (risk === 'liability_limit') {
    if (/\bliability\b|\bliable\b/.test(lower)) {
      return 'liability';
    }
    if (/\bwarrant(?:y|ies)\b|\bas\s+is\b/.test(lower)) {
      return 'warranty';
    }
    if (/\bdamages?\b|\bloss(?:es)?\b/.test(lower)) {
      return 'damages';
    }
  }
  if (risk === 'mandatory_arbitration') {
    return 'dispute';
  }
  if (risk === 'jurisdiction_governing_law') {
    if (/\blaw\b|\blaws\s+of\b|\bgoverned\s+by\b/.test(lower)) {
      return 'law';
    }
    return 'jurisdiction';
  }
  if (risk === 'content_license') {
    if (/\bfeedback\b|\bsuggestion\b/.test(lower)) {
      return 'feedback';
    }
    return /\buser\s+content\b|\byour\s+content\b|\bcontent\b/.test(lower)
      ? 'user_content'
      : 'unknown';
  }
  if (risk === 'data_or_privacy') {
    if (/\bpersonal\s+data\b|\bpersonal\s+information\b|\byour\s+data\b|\binformation\b/.test(lower)) {
      return 'personal_data';
    }
  }
  if (/\bterms\b|\bagreement\b|\bcontract\b/.test(lower)) {
    return 'terms';
  }
  if (/\bfees?\b|\bprice\b|\bcharge\b|\bbilling\b|\bsubscription\b/.test(lower)) {
    return 'fees';
  }
  if (/\bservice\b|\bofferings?\b|\bfeatures?\b/.test(lower)) {
    return 'service';
  }
  if (/\baccount\b|\baccess\b/.test(lower)) {
    return 'account';
  }
  if (/\buser\s+content\b|\byour\s+content\b|\bcontent\b/.test(lower)) {
    return risk === 'data_or_privacy' ? 'personal_data' : 'user_content';
  }
  if (/\bpersonal\s+data\b|\bpersonal\s+information\b|\byour\s+data\b|\binformation\b/.test(lower)) {
    return 'personal_data';
  }
  if (/\bliability\b|\bliable\b/.test(lower)) {
    return 'liability';
  }
  if (/\bwarrant(?:y|ies)\b|\bas\s+is\b/.test(lower)) {
    return 'warranty';
  }
  if (/\bdamages?\b|\bloss(?:es)?\b/.test(lower)) {
    return 'damages';
  }
  if (/\bdispute\b|\bclaim\b|\barbitration\b|\bclass\s+action\b|\bjury\s+trial\b/.test(lower)) {
    return 'dispute';
  }
  if (/\bjurisdiction\b|\bvenue\b|\bcourts?\b|\bforum\b/.test(lower)) {
    return 'jurisdiction';
  }
  if (/\blaw\b|\blaws\s+of\b|\bgoverned\s+by\b/.test(lower)) {
    return 'law';
  }
  if (/\bprivacy\s+policy\b|\buser\s+guidelines\b|\bother\s+terms\b|\bdocument\b/.test(lower)) {
    return 'document';
  }
  if (/\bfeedback\b|\bsuggestion\b/.test(lower)) {
    return 'feedback';
  }
  return 'unknown';
}

function inferScope(text: string, risk: LegaleseRiskKind): string[] {
  const lower = text.toLowerCase();
  const scope: string[] = [];
  if (/\bany\s+time\b|\bfrom\s+time\s+to\s+time\b/.test(lower)) {
    scope.push('any_time');
  }
  if (/\bwithout\s+(?:prior\s+)?notice\b|\bno\s+notice\b/.test(lower)) {
    scope.push('without_notice');
  }
  if (/\bnotice\b|\bnotify\b|\bcommunicate\b|\bposting\b/.test(lower)) {
    scope.push('notice');
  }
  if (/\bworldwide\b/.test(lower)) {
    scope.push('worldwide');
  }
  if (/\broyalty[-\s]?free\b|\bfully\s+paid\b/.test(lower)) {
    scope.push('royalty_free');
  }
  if (/\bsublicens(?:e|able)\b/.test(lower)) {
    scope.push('sublicensable');
  }
  if (/\btransferable\b/.test(lower)) {
    scope.push('transferable');
  }
  if (/\bperpetual\b|\birrevocable\b/.test(lower)) {
    scope.push('perpetual');
  }
  if (/\bnon[-\s]?exclusive\b/.test(lower)) {
    scope.push('nonexclusive');
  }
  if (/\brecurring\b|\bauto[-\s]?renew\b/.test(lower)) {
    scope.push('recurring');
  }
  if (risk === 'jurisdiction_governing_law' && /\bconsumer\b|\bhabitual\s+residence\b/.test(lower)) {
    scope.push('consumer_law_context');
  }
  return uniqueStrings(scope);
}

function inferConditions(text: string): string[] {
  const lower = text.toLowerCase();
  const conditions: string[] = [];
  if (/\bnotice\b|\bnotify\b|\bcommunicate\b|\bposting\b/.test(lower)) {
    conditions.push('notice');
  }
  if (/\bserious\b|\bcause\b|\bviolation\b|\bbreach\b|\breasonable\b/.test(lower)) {
    conditions.push('serious_grounds');
  }
  if (/\bapplicable\s+law\b|\bpermitted\s+by\s+law\b|\bto\s+the\s+extent\b/.test(lower)) {
    conditions.push('applicable_law');
  }
  if (/\bbusiness\b|\benterprise\b|\bcommercial\b/.test(lower)) {
    conditions.push('business_only');
  }
  if (/\bconsent\b|\bagree\b|\bapproval\b/.test(lower)) {
    conditions.push('consent');
  }
  if (/\b\d+\s+(?:days?|months?|years?)\b|\bafter\s+\d+\b/.test(lower)) {
    conditions.push('time_period');
  }
  return uniqueStrings(conditions);
}

function inferExceptions(text: string): string[] {
  const lower = text.toLowerCase();
  const exceptions: string[] = [];
  if (/\bexcept\b|\bexception\b/.test(lower)) {
    exceptions.push('explicit_exception');
  }
  if (/\bapplicable\s+law\b|\bpermitted\s+by\s+law\b|\bto\s+the\s+extent\b/.test(lower)) {
    exceptions.push('applicable_law');
  }
  if (/\bconsumer\b|\bstatutory\b/.test(lower)) {
    exceptions.push('statutory_rights');
  }
  return uniqueStrings(exceptions);
}

function matchEvidenceSpan(text: string, startOffset: number, risk: LegaleseRiskKind): LegaleseSpan {
  const pattern = evidencePattern(risk);
  const match = pattern.exec(text);
  if (!match || match.index < 0) {
    return {
      kind: 'frame',
      startOffset,
      endOffset: startOffset + text.length,
      text,
    };
  }
  const matchedText = match[0] ?? text;
  return {
    kind: 'frame',
    startOffset: startOffset + match.index,
    endOffset: startOffset + match.index + matchedText.length,
    text: matchedText,
  };
}

function evidencePattern(risk: LegaleseRiskKind): RegExp {
  switch (risk) {
    case 'unilateral_change':
      return /\b(change|modify|update|revise)\b.{0,80}\b(terms|agreement|service|fees|account|content)\b/i;
    case 'termination_or_suspension':
      return /\b(terminate|suspend|disable|delete|refuse)\b.{0,80}\b(account|access|service|content)?/i;
    case 'liability_limit':
      return /\b(no|not)\s+liab(?:le|ility)\b|\blimitation\s+of\s+liability\b|\bwithout\s+warrant(?:y|ies)\b|\bas\s+is\b/i;
    case 'mandatory_arbitration':
      return /\barbitration\b|\bclass\s+action\b|\bjury\s+trial\b|\bcollective\s+action\b/i;
    case 'jurisdiction_governing_law':
      return /\bgoverning\s+law\b|\bjurisdiction\b|\bvenue\b|\blaws\s+of\b|\bcourts?\s+of\b/i;
    case 'content_license':
      return /\b(user\s+content|license|royalty[-\s]?free|worldwide|sublicens(?:e|able))\b/i;
    case 'payment_or_auto_renewal':
      return /\b(subscription|auto[-\s]?renew|billing|refund|fee|charge|paid)\b/i;
    case 'data_or_privacy':
      return /\b(personal\s+data|personal\s+information|privacy\s+policy|collect|process(?:ing)?|share)\b/i;
    case 'third_party_dependency':
      return /\b(third[-\s]?party|affiliate|partner|external\s+link|privacy\s+policy|user\s+guidelines)\b/i;
    case 'unclear_language':
      return /\b(applicable\s+law|to\s+the\s+extent|some\s+jurisdictions)\b/i;
  }
}

function riskConfidenceFromText(text: string, risk: LegaleseRiskKind): number {
  const rule = RISK_RULES.find((entry) => entry.kind === risk);
  if (rule?.pattern.test(text)) {
    return rule.confidence;
  }
  return 0.5;
}

function extractDefinitions(
  clauseId: string,
  text: string,
  startOffset: number
): LegaleseDefinition[] {
  const definitions: LegaleseDefinition[] = [];
  const patterns = [
    /"([^"]{2,80})"\s+(?:means|shall\s+mean|refers\s+to|is\s+defined\s+as)\s+([^.;]+)/gi,
    /\b([A-Z][A-Za-z0-9 ]{2,80})\s+(?:means|shall\s+mean|refers\s+to|is\s+defined\s+as)\s+([^.;]+)/g,
  ];

  for (const pattern of patterns) {
    let match = pattern.exec(text);
    while (match) {
      const term = sanitizeInlineText(match[1] ?? '');
      const meaning = sanitizeInlineText(match[2] ?? '');
      const fullText = match[0] ?? '';
      if (term && meaning) {
        definitions.push({
          id: `${clauseId}_definition_${definitions.length + 1}`,
          term,
          meaning,
          span: {
            kind: 'definition',
            startOffset: startOffset + match.index,
            endOffset: startOffset + match.index + fullText.length,
            text: fullText,
          },
        });
      }
      match = pattern.exec(text);
    }
  }

  return definitions;
}

function extractCrossReferences(
  clauseId: string,
  text: string,
  startOffset: number,
  sections: LegaleseSection[]
): LegaleseCrossReference[] {
  const references: LegaleseCrossReference[] = [];
  const patterns = [
    /\b(?:Section|Sections|Clause|Clauses)\s+(\d+(?:\.\d+)*)\b/gi,
    /\b(Privacy\s+Policy|User\s+Guidelines|Card\s+Terms|these\s+Terms|this\s+Agreement|other\s+terms)\b/gi,
  ];

  for (const pattern of patterns) {
    let match = pattern.exec(text);
    while (match) {
      const rawText = match[0] ?? '';
      const target = sanitizeInlineText(match[1] ?? rawText);
      const resolvedSection = resolveSectionReference(target, sections);
      const reference: LegaleseCrossReference = {
        id: `${clauseId}_xref_${references.length + 1}`,
        rawText,
        target,
        span: {
          kind: 'cross_reference',
          startOffset: startOffset + match.index,
          endOffset: startOffset + match.index + rawText.length,
          text: rawText,
        },
        resolved: Boolean(resolvedSection),
      };
      if (resolvedSection) {
        reference.resolvedSectionId = resolvedSection.id;
        reference.resolvedSectionPath = resolvedSection.path;
      }
      references.push(reference);
      match = pattern.exec(text);
    }
  }

  return references;
}

function resolveSectionReference(
  target: string,
  sections: LegaleseSection[]
): LegaleseSection | undefined {
  const normalized = target.toLowerCase().replace(/\s+/g, ' ').trim();
  return sections.find((section) => {
    const path = section.path.join('.');
    return path === normalized || section.heading.toLowerCase() === normalized;
  });
}

function shouldRequireResolvedReference(rawText: string): boolean {
  return /\b(section|clause)\b/i.test(rawText);
}

function emitLegaleseGg(
  documentId: string,
  serviceName: string,
  options: LegaleseLoweringOptions,
  sections: LegaleseSection[],
  clauses: LegaleseClauseLowering[],
  risks: LegaleseRiskKind[],
  summary: LegaleseLoweringSummary
): string {
  const summaryId = `${documentId}_summary`;
  const measureId = `${documentId}_measure`;
  const lines = [
    '// legalese-to-gg lowering v2',
    `(${documentId}:LegalDocument { language: "legalese", lowering: "legalese-v2", service: ${ggString(
      serviceName
    )}, source: ${ggString(options.sourcePath ?? 'inline')}, clauses: "${
      clauses.length
    }", sections: "${sections.length}" })`,
    `(${summaryId}:LegalSummary { clause_count: "${clauses.length}", risk_count: "${
      risks.length
    }", hard_clauses: "${summary.hardClauseCount}", opaque_clauses: "${
      summary.opaqueClauseCount
    }", avg_words_per_clause: "${summary.averageWordsPerClause.toFixed(2)}" })`,
    `(${measureId}:LegalMeasure { metric: "tos_validation", risk_count: "${
      risks.length
    }" })`,
  ];

  for (const section of sections) {
    lines.push(
      `(${sectionNodeId(documentId, section)}:LegalSection { index: "${
        section.index + 1
      }", path: ${ggString(section.path.join('.'))}, heading: ${ggString(
        section.heading
      )}, start: "${section.startOffset}", end: "${section.endOffset}" })`
    );
    lines.push(`(${documentId})-[:CONTAINS_SECTION]->(${sectionNodeId(documentId, section)})`);
  }

  for (const risk of risks) {
    lines.push(
      `(${riskNodeId(documentId, risk)}:LegalRisk { kind: "${risk}", label: ${ggString(
        riskLabel(risk)
      )} })`
    );
  }

  for (const clause of clauses) {
    lines.push(
      `(${clause.id}:LegalClause { index: "${clause.index + 1}", kind: "${
        clause.kind
      }", actor: "${clause.actor}", clarity: "${clause.clarity}", section_path: ${ggString(
        clause.sectionPath.join('.')
      )}, heading: ${ggString(clause.heading)}, words: "${clause.wordCount}", sentences: "${
        clause.sentenceCount
      }", legal_terms: "${clause.legalTermCount}", void_boundary: "${
        clause.risks.length
      }", holes: "${clause.holes.length}", excerpt: ${ggString(clause.excerpt)} })`
    );
    const section = findSectionForClause(clause, sections);
    if (section) {
      lines.push(`(${sectionNodeId(documentId, section)})-[:CONTAINS_CLAUSE]->(${clause.id})`);
    }
  }

  if (clauses.length === 0) {
    lines.push(`(${documentId})-[:PROCESS { lowering: "legalese-v2" }]->(${summaryId})`);
  } else if (clauses.length === 1) {
    const onlyClause = clauses[0];
    if (onlyClause === undefined) {
      throw new Error('Expected a single legalese clause but none was available.');
    }
    lines.push(
      `(${documentId})-[:PROCESS { lowering: "legalese-v2" }]->(${onlyClause.id})`
    );
    lines.push(`(${onlyClause.id})-[:PROCESS { role: "summarize" }]->(${summaryId})`);
  } else {
    lines.push(
      `(${documentId})-[:FORK { lowering: "legalese-v2", void_density: "${riskDensity(
        clauses
      ).toFixed(3)}" }]->(${clauses.map((clause) => clause.id).join('|')})`
    );
    lines.push(
      `(${clauses
        .map((clause) => clause.id)
        .join('|')})-[:FOLD { strategy: "preserve-multiplicity", evidence: "clause-frame-signals" }]->(${summaryId})`
    );
  }

  for (const clause of clauses) {
    emitClauseGg(lines, documentId, clause);
  }

  lines.push(
    `(${summaryId})-[:MEASURE { metric: "legalese_risk", value: "${
      risks.length
    }" }]->(${measureId})`
  );

  return `${lines.join('\n')}\n`;
}

function emitClauseGg(lines: string[], documentId: string, clause: LegaleseClauseLowering): void {
  for (const risk of clause.risks) {
    lines.push(
      `(${clause.id})-[:OBSERVE { risk: "${risk}", confidence: "${riskConfidence(
        clause,
        risk
      ).toFixed(2)}" }]->(${riskNodeId(documentId, risk)})`
    );
  }

  for (const definition of clause.definitions) {
    lines.push(
      `(${definition.id}:LegalDefinition { term: ${ggString(definition.term)}, meaning: ${ggString(
        definition.meaning
      )}, start: "${definition.span.startOffset}", end: "${definition.span.endOffset}" })`
    );
    lines.push(`(${clause.id})-[:DEFINED_BY]->(${definition.id})`);
  }

  for (const reference of clause.crossReferences) {
    lines.push(
      `(${reference.id}:LegalCrossReference { raw: ${ggString(
        reference.rawText
      )}, target: ${ggString(reference.target)}, resolved: "${reference.resolved}" })`
    );
    lines.push(`(${clause.id})-[:REFERENCES]->(${reference.id})`);
  }

  for (const frame of clause.frames) {
    emitFrameGg(lines, documentId, clause, frame);
  }

  for (const hole of clause.holes) {
    lines.push(
      `(${hole.id}:LegalFrameHole { kind: "${hole.kind}", message: ${ggString(
        hole.message
      )}, start: "${hole.span.startOffset}", end: "${hole.span.endOffset}" })`
    );
    lines.push(`(${clause.id})-[:REQUIRES_PROOF]->(${hole.id})`);
  }
}

function emitFrameGg(
  lines: string[],
  documentId: string,
  clause: LegaleseClauseLowering,
  frame: LegaleseClauseFrame
): void {
  const actorId = `${frame.id}_actor_${frame.actor}`;
  const recipientId = `${frame.id}_recipient_${frame.recipient}`;
  const actionId = `${frame.id}_action_${frame.action}`;
  const objectId = `${frame.id}_object_${frame.object}`;
  lines.push(
    `(${frame.id}:LegalClauseFrame { risk: "${frame.riskKind}", operator: "${
      frame.operator
    }", action: "${frame.action}", object: "${frame.object}", confidence: "${frame.confidence.toFixed(
      2
    )}", scope: ${ggString(frame.scope.join(','))} })`
  );
  lines.push(`(${actorId}:LegalParty { role: "${frame.actor}" })`);
  lines.push(`(${recipientId}:LegalParty { role: "${frame.recipient}" })`);
  lines.push(`(${actionId}:LegalAction { kind: "${frame.action}", modality: "${frame.operator}" })`);
  lines.push(`(${objectId}:LegalObject { kind: "${frame.object}" })`);
  lines.push(`(${clause.id})-[:ASSERTS]->(${frame.id})`);
  lines.push(`(${frame.id})-[:BIND_PARTY { role: "actor" }]->(${actorId})`);
  lines.push(`(${frame.id})-[:BIND_PARTY { role: "recipient" }]->(${recipientId})`);
  lines.push(`(${frame.id})-[:MODALITY]->(${actionId})`);
  lines.push(`(${frame.id})-[:ACTS_ON]->(${objectId})`);
  lines.push(`(${frame.id})-[:CREATES_RISK]->(${riskNodeId(documentId, frame.riskKind)})`);

  for (const [index, condition] of frame.conditions.entries()) {
    const conditionId = `${frame.id}_condition_${index + 1}`;
    lines.push(`(${conditionId}:LegalCondition { kind: ${ggString(condition)} })`);
    lines.push(`(${frame.id})-[:CONDITIONED_BY]->(${conditionId})`);
  }
  for (const [index, exception] of frame.exceptions.entries()) {
    const exceptionId = `${frame.id}_exception_${index + 1}`;
    lines.push(`(${exceptionId}:LegalException { kind: ${ggString(exception)} })`);
    lines.push(`(${frame.id})-[:EXCEPTED_BY]->(${exceptionId})`);
  }
}

function buildSummary(
  documentId: string,
  serviceName: string,
  clauses: LegaleseClauseLowering[]
): LegaleseLoweringSummary {
  const riskCounts = Object.fromEntries(ALL_RISK_KINDS.map((risk) => [risk, 0])) as Record<
    LegaleseRiskKind,
    number
  >;
  let wordTotal = 0;
  let hardClauseCount = 0;
  let opaqueClauseCount = 0;

  for (const clause of clauses) {
    wordTotal += clause.wordCount;
    if (clause.clarity === 'hard') {
      hardClauseCount++;
    } else if (clause.clarity === 'opaque') {
      opaqueClauseCount++;
    }
    for (const risk of clause.risks) {
      riskCounts[risk]++;
    }
  }

  return {
    documentId,
    serviceName,
    clauseCount: clauses.length,
    riskCounts,
    hardClauseCount,
    opaqueClauseCount,
    averageWordsPerClause: clauses.length === 0 ? 0 : wordTotal / clauses.length,
  };
}

function normalizeLegaleseSource(input: string): string {
  return input
    .replace(/\r\n?/g, '\n')
    .replace(/\f/g, '\n\n')
    .replace(/[ \t]+\n/g, '\n')
    .trim();
}

function splitLinesWithOffsets(text: string): SourceLine[] {
  const lines: SourceLine[] = [];
  let startOffset = 0;
  for (let index = 0; index <= text.length; index++) {
    if (index === text.length || text[index] === '\n') {
      lines.push({
        text: text.slice(startOffset, index),
        startOffset,
        endOffset: index,
      });
      startOffset = index + 1;
    }
  }
  return lines;
}

function spanFromLines(lines: SourceLine[]): SourceSpan {
  const first = lines[0];
  const last = lines[lines.length - 1];
  if (!first || !last) {
    return { text: '', startOffset: 0, endOffset: 0 };
  }
  return {
    text: lines.map((line) => line.text).join('\n'),
    startOffset: first.startOffset,
    endOffset: last.endOffset,
  };
}

function parseHeadingBlock(
  text: string,
  lines: SourceLine[]
): { path: string[]; heading: string } | null {
  if (lines.length !== 1) {
    return null;
  }
  const trimmed = text.trim();
  if (trimmed.length < 3 || trimmed.length > 140) {
    return null;
  }
  if (/[.;!?]$/.test(trimmed) || /\b(may|must|shall|will|agree|liable|arbitration|license)\b/i.test(trimmed)) {
    return null;
  }

  const numbered = /^(?:section\s+)?(\d+(?:\.\d+)*)(?:[.)])?\s*(.+)$/i.exec(trimmed);
  if (numbered) {
    return {
      path: (numbered[1] ?? '').split('.').filter((part) => part.length > 0),
      heading: sanitizeInlineText(numbered[2] ?? trimmed),
    };
  }

  const allCaps = trimmed === trimmed.toUpperCase() && /[A-Z]/.test(trimmed);
  const titleLike = /^[A-Z][A-Za-z0-9 ,/&'()-]+$/.test(trimmed);
  if (allCaps || titleLike) {
    return {
      path: [toNodeId(trimmed, 'section')],
      heading: sanitizeInlineText(trimmed),
    };
  }

  return null;
}

function splitBlockIntoCandidateSpans(blockSpan: SourceSpan, lines: SourceLine[]): SourceSpan[] {
  const nonBlankLines = lines.filter((line) => line.text.trim().length > 0);
  if (
    nonBlankLines.length > 1 &&
    nonBlankLines.every((line) => /^\s*(?:[-*]|\d+[.)]|\([a-z0-9]+\))\s+/i.test(line.text))
  ) {
    return nonBlankLines.map((line) => trimSourceSpan(line));
  }
  return [blockSpan];
}

function splitSentences(block: string): string[] {
  return splitSentencesWithOffsets({ text: block, startOffset: 0, endOffset: block.length }).map(
    (sentence) => normalizeClauseText(sentence.text)
  );
}

function splitSentencesWithOffsets(block: SourceSpan): SourceSpan[] {
  const sentences: SourceSpan[] = [];
  let start = 0;

  for (let index = 0; index < block.text.length; index++) {
    const char = block.text[index];
    if (char !== '.' && char !== ';' && char !== '?' && char !== '!') {
      continue;
    }

    const nextStart = findNextNonSpace(block.text, index + 1);
    if (nextStart >= block.text.length) {
      sentences.push({
        text: block.text.slice(start).trim(),
        startOffset: block.startOffset + start,
        endOffset: block.startOffset + block.text.length,
      });
      start = block.text.length;
      break;
    }

    const nextChar = block.text[nextStart] ?? '';
    if (/^[A-Z0-9(-]/.test(nextChar)) {
      sentences.push({
        text: block.text.slice(start, index + 1).trim(),
        startOffset: block.startOffset + start,
        endOffset: block.startOffset + index + 1,
      });
      start = nextStart;
    }
  }

  if (start < block.text.length) {
    sentences.push({
      text: block.text.slice(start).trim(),
      startOffset: block.startOffset + start,
      endOffset: block.endOffset,
    });
  }

  return sentences.filter((sentence) => sentence.text.length > 0);
}

function splitOversizedClauseWithOffsets(text: SourceSpan, maxChars: number): SourceSpan[] {
  if (text.text.length <= maxChars) {
    return [text];
  }

  const clauses: SourceSpan[] = [];
  let rest = text.text.trim();
  let restOffset = text.startOffset + Math.max(0, text.text.indexOf(rest));
  while (rest.length > maxChars) {
    const cut = findClauseCut(rest, maxChars);
    clauses.push({
      text: rest.slice(0, cut).trim(),
      startOffset: restOffset,
      endOffset: restOffset + cut,
    });
    rest = rest.slice(cut).trim();
    restOffset += cut;
  }
  if (rest.length > 0) {
    clauses.push({
      text: rest,
      startOffset: restOffset,
      endOffset: restOffset + rest.length,
    });
  }
  return clauses;
}

function trimSourceSpan(span: SourceSpan): SourceSpan {
  const leading = span.text.match(/^\s*/)?.[0].length ?? 0;
  const trailing = span.text.match(/\s*$/)?.[0].length ?? 0;
  return {
    text: span.text.slice(leading, span.text.length - trailing),
    startOffset: span.startOffset + leading,
    endOffset: span.endOffset - trailing,
  };
}

function normalizeClauseText(text: string): string {
  return text.replace(/\s*\n\s*/g, ' ').replace(/\s+/g, ' ').trim();
}

function findClauseCut(text: string, maxChars: number): number {
  const windowStart = Math.max(0, maxChars - 240);
  const slice = text.slice(windowStart, maxChars);
  const lastBreak = Math.max(slice.lastIndexOf(';'), slice.lastIndexOf(','), slice.lastIndexOf(' and '));
  if (lastBreak > 0) {
    return windowStart + lastBreak + 1;
  }
  const lastSpace = text.lastIndexOf(' ', maxChars);
  return lastSpace > 0 ? lastSpace : maxChars;
}

function findNextNonSpace(text: string, start: number): number {
  for (let index = start; index < text.length; index++) {
    if (!/\s/.test(text[index] ?? '')) {
      return index;
    }
  }
  return text.length;
}

function tokenizeWords(text: string): string[] {
  return text.match(/[A-Za-z0-9]+(?:[-'][A-Za-z0-9]+)?/g) ?? [];
}

function countLegalTerms(words: string[]): number {
  const legalTerms = new Set(LEGAL_TERMS);
  return words.filter((word) => legalTerms.has(word.toLowerCase())).length;
}

function scoreClarity(
  text: string,
  words: string[],
  averageWordsPerSentence: number,
  legalTermCount: number
): number {
  const longWordCount = words.filter((word) => word.length >= 12).length;
  const passiveCount = (text.match(/\b(?:is|are|be|been|being|was|were)\s+\w+ed\b/gi) ?? [])
    .length;
  return (
    averageWordsPerSentence / 24 +
    legalTermCount / Math.max(1, words.length) * 8 +
    longWordCount / Math.max(1, words.length) * 5 +
    passiveCount * 0.35
  );
}

function classifyClarity(score: number): LegaleseClarity {
  if (score >= 3.1) {
    return 'opaque';
  }
  if (score >= 2.1) {
    return 'hard';
  }
  if (score >= 1.25) {
    return 'moderate';
  }
  return 'plain';
}

function inferActor(text: string): LegaleseActor {
  const lower = text.toLowerCase();
  const hasUser = /\b(you|user|users|customer|subscriber)\b/.test(lower);
  const hasProvider = /\b(we|us|our|company|provider|service provider|platform|fintecom)\b/.test(lower);
  if (hasUser && hasProvider) {
    return 'both';
  }
  if (hasUser) {
    return 'user';
  }
  if (hasProvider) {
    return 'provider';
  }
  if (/\bthird[-\s]?party|affiliate|partner\b/.test(lower)) {
    return 'third_party';
  }
  return 'unknown';
}

function inferServiceName(input: string): string {
  const firstLine = input
    .split(/\r?\n/)
    .map((line) => line.trim())
    .find((line) => line.length >= 3 && line.length <= 120);
  if (!firstLine) {
    return 'Terms of Service';
  }
  return firstLine.replace(/\s+/g, ' ');
}

function uniqueRiskKinds(risks: LegaleseRiskKind[]): LegaleseRiskKind[] {
  return [...new Set(risks)];
}

function uniqueStrings(values: string[]): string[] {
  return [...new Set(values)];
}

function riskDensity(clauses: LegaleseClauseLowering[]): number {
  if (clauses.length === 0) {
    return 0;
  }
  const riskyClauses = clauses.filter((clause) => clause.risks.length > 0).length;
  return riskyClauses / clauses.length;
}

function riskConfidence(clause: LegaleseClauseLowering, risk: LegaleseRiskKind): number {
  const signal = clause.signals.find((entry) => entry.kind === risk);
  const frame = clause.frames.find((entry) => entry.riskKind === risk);
  return signal?.confidence ?? frame?.confidence ?? 0.5;
}

function findSectionForClause(
  clause: LegaleseClauseLowering,
  sections: LegaleseSection[]
): LegaleseSection | undefined {
  const path = clause.sectionPath.join('.');
  return sections.find((section) => section.path.join('.') === path);
}

function sectionNodeId(documentId: string, section: LegaleseSection): string {
  return `${documentId}_${section.id}`.replace(/[^A-Za-z0-9_]/g, '_');
}

function riskNodeId(documentId: string, risk: LegaleseRiskKind): string {
  return `${documentId}_risk_${risk}`;
}

function riskLabel(risk: LegaleseRiskKind): string {
  return risk.replace(/_/g, ' ');
}

function toNodeId(value: string, fallback: string): string {
  const normalized = value
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '_')
    .replace(/^_+|_+$/g, '')
    .slice(0, 72);
  return normalized || fallback;
}

function sanitizePropertyText(value: string, maxChars: number): string {
  return value
    .normalize('NFKD')
    .replace(/[^\x20-\x7E]/g, '')
    .replace(/["'{}()[\]|\\]/g, ' ')
    .replace(/\s+/g, ' ')
    .trim()
    .slice(0, maxChars);
}

function sanitizeInlineText(value: string): string {
  return value
    .normalize('NFKD')
    .replace(/[^\x20-\x7E]/g, '')
    .replace(/\s+/g, ' ')
    .trim();
}

function ggString(value: string): string {
  return `"${sanitizePropertyText(value, Math.max(DEFAULT_EXCERPT_CHARS, value.length))}"`;
}
