// @a0n/.ignored_gnosis/src/betty/mudras.ts
/**
 * Mudras Linter: Geometric Crystallography Static Analysis
 *
 * Traditional static analysis uses control flow graphs to trace time-bound logic.
 * Mudras operates on the Amplituhedron, where all states exist simultaneously as
 * facets of a multidimensional shape. It detects topological and geometric fractures.
 */

import type { GraphAST } from './compiler';
import type { AmplituhedronGeometry } from '../betti/amplituhedron-volume';

export interface MudrasDiagnostic {
  type:
    | 'PositivityViolation'
    | 'TypeMismatch'
    | 'DeadCode'
    | 'CohomologicalDeadlock';
  message: string;
}

export interface MudrasResult {
  diagnostics: MudrasDiagnostic[];
  optimizedGeometry: AmplituhedronGeometry;
}

/**
 * Executes the Mudras static analysis pipeline on a Grassmannian shape.
 */
export function runMudrasCrystallography(
  ast: GraphAST,
  geometry: AmplituhedronGeometry
): MudrasResult {
  const diagnostics: MudrasDiagnostic[] = [];
  const optimizedGeometry = { ...geometry };

  // PHASE 1: Plücker Coordinate Positivity (Memory Safety)
  // The shape must reside strictly within the Positive Grassmannian.
  if (
    optimizedGeometry.dimensionality <= 0 ||
    optimizedGeometry.staticVolume <= 0
  ) {
    diagnostics.push({
      type: 'PositivityViolation',
      message: `Plücker coordinate failure: Dimensionality (${optimizedGeometry.dimensionality}) or Volume (${optimizedGeometry.staticVolume}) <= 0. The logic has pierced the hull of reality into negative space.`,
    });
  }

  // PHASE 2: Schubert Calculus (Type Checking)
  // Ensures all sub-graphs intersect (data can flow). If there are disjoint components,
  // the Schubert intersection number is exactly 0.
  const isDisjoint = checkDisjointSubgraphs(ast);
  if (isDisjoint) {
    diagnostics.push({
      type: 'TypeMismatch',
      message: `Schubert Calculus failure: Intersection volume is exactly 0. Two components of the topology do not share physical geometry and cannot pass data.`,
    });
  }

  // PHASE 3: Boundary Stratification (Dead Code Elimination)
  // Detects "soft" boundaries that contribute 0 volume (e.g. disconnected sink nodes).
  const softBoundaries = detectSoftBoundaries(ast);
  if (softBoundaries > 0) {
    diagnostics.push({
      type: 'DeadCode',
      message: `Boundary Stratification: Detected ${softBoundaries} soft boundaries contributing zero volume. Shearing facets to optimize shape.`,
    });
    // Optimization: Shear the soft boundaries off the geometry
    optimizedGeometry.staticVolume -=
      softBoundaries * optimizedGeometry.dimensionality;
  }

  // PHASE 4: Cohomological Obstructions (Deadlock Detection)
  // Detects closed topological loops (infinite loops wrapping around empty space).
  // In cohomology, this is a non-zero 1st Betti number (β1 > 0).
  const betti1 = calculateFirstBettiNumber(ast);
  if (betti1 > 0) {
    diagnostics.push({
      type: 'CohomologicalDeadlock',
      message: `Cohomological Obstruction: 1st Betti Number (β1) = ${betti1}. The shape contains a topological hole (infinite loop), trapping empty space.`,
    });
  }

  return { diagnostics, optimizedGeometry };
}

/**
 * Calculates if there are disconnected components in the AST.
 */
function checkDisjointSubgraphs(ast: GraphAST): boolean {
  if (ast.nodes.size === 0) return false;

  const visited = new Set<string>();
  const nodes = Array.from(ast.nodes.values());
  const startNode = nodes[0]!.id;

  // Build adjacency list
  const adj = new Map<string, string[]>();
  for (const n of ast.nodes.keys()) adj.set(n, []);

  for (const edge of ast.edges) {
    for (const src of edge.sourceIds) {
      for (const tgt of edge.targetIds) {
        adj.get(src)?.push(tgt);
        adj.get(tgt)?.push(src); // undirected for connectivity check
      }
    }
  }

  // DFS
  const stack = [startNode];
  while (stack.length > 0) {
    const node = stack.pop()!;
    if (!visited.has(node)) {
      visited.add(node);
      const neighbors = adj.get(node) || [];
      for (const neighbor of neighbors) {
        stack.push(neighbor);
      }
    }
  }

  // If we couldn't visit all nodes from the start node, the graph is disjoint.
  return visited.size < ast.nodes.size;
}

/**
 * Detects soft boundaries (e.g. nodes with no incoming edges that aren't the clinamen).
 */
function detectSoftBoundaries(ast: GraphAST): number {
  let softFaces = 0;
  const hasIncoming = new Set<string>();

  for (const edge of ast.edges) {
    for (const tgt of edge.targetIds) {
      hasIncoming.add(tgt);
    }
  }

  for (const node of ast.nodes.values()) {
    // If a node has no incoming edges and is not explicitly labeled as the start/Clinamen
    if (!hasIncoming.has(node.id) && !node.labels.includes('Clinamen')) {
      softFaces++;
    }
  }

  return softFaces;
}

/**
 * Calculates the first Betti number (β1 = E - V + C)
 */
function calculateFirstBettiNumber(ast: GraphAST): number {
  const V = ast.nodes.size;
  const E = ast.edges.reduce(
    (acc, edge) => acc + edge.sourceIds.length * edge.targetIds.length,
    0
  );

  // Assuming 1 connected component for a valid topology
  const C = checkDisjointSubgraphs(ast) ? 2 : 1;

  const betti1 = E - V + C;
  return Math.max(0, betti1);
}
