// @a0n/.ignored_gnosis/src/betty/p-adic-topology.ts
/**
 * The p-Adic Topology Substrate
 * 
 * Abandons the Archimedean Axiom. Distance is no longer measured by physical 
 * magnitude along a line, but by divisibility by a prime base (p).
 * 
 * Enforces the Strong Triangle Inequality:
 * d(x, z) <= max(d(x, y), d(y, z))
 */

export const PRIME_BASE = 2; // p = 2 for memory topology

/**
 * Calculates the p-adic valuation (highest power of p that divides n).
 */
export function pAdicValuation(n: number, p: number = PRIME_BASE): number {
  if (n === 0) return Infinity; // Valuation of 0 is infinity
  
  let valuation = 0;
  let num = Math.abs(n);
  
  // Fractional numbers need rational expansion, but for memory indexing
  // we deal with integers.
  if (!Number.isInteger(num)) {
    throw new Error("Topological constraint: memory addresses must be integers.");
  }

  while (num % p === 0) {
    valuation++;
    num /= p;
  }
  
  return valuation;
}

/**
 * Calculates the p-adic norm (absolute value in p-adic space).
 * |x|_p = p^(-v_p(x))
 */
export function pAdicNorm(n: number, p: number = PRIME_BASE): number {
  if (n === 0) return 0;
  return Math.pow(p, -pAdicValuation(n, p));
}

/**
 * Calculates the Ultrametric Distance between two points.
 * d(x, y) = |x - y|_p
 */
export function ultrametricDistance(x: number, y: number, p: number = PRIME_BASE): number {
  return pAdicNorm(x - y, p);
}

/**
 * An Ultrametric Memory Sphere in p-adic space.
 * S(center, radius) = { x : d(x, center) <= radius }
 */
export class UltrametricSphere {
  public center: number;
  public radius: number;
  public p: number;

  constructor(center: number, radius: number, p: number = PRIME_BASE) {
    this.center = center;
    this.radius = radius;
    this.p = p;
  }

  /**
   * Checks if an address point exists within the memory sphere.
   */
  public contains(x: number): boolean {
    return ultrametricDistance(x, this.center, this.p) <= this.radius;
  }

  /**
   * Evaluates collision under the Strong Triangle Inequality.
   * In p-adic space, two spheres NEVER partially intersect. 
   * They are either disjoint, or one is completely contained in the other.
   */
  public getIntersectionState(other: UltrametricSphere): 'DISJOINT' | 'ENCAPSULATED_BY' | 'ENCAPSULATES' | 'IDENTICAL' {
    if (this.p !== other.p) throw new Error("Cannot evaluate spheres across different prime topologies.");

    const distance = ultrametricDistance(this.center, other.center, this.p);
    
    // If distance is greater than both radii, they are completely disjoint.
    if (distance > Math.max(this.radius, other.radius)) {
      return 'DISJOINT';
    }

    // If distance is less than or equal to the larger radius, the larger sphere ENTIRELY contains the smaller.
    if (this.radius === other.radius) {
      return 'IDENTICAL';
    } else if (this.radius > other.radius) {
      return 'ENCAPSULATES';
    } else {
      return 'ENCAPSULATED_BY';
    }
  }
}
