// @a0n/.ignored_gnosis/src/betty/semantic-validator.ts
/**
 * Runtime semantic type validator.
 *
 * Validates JSON values against TopologyType contracts at runtime.
 * Used by the PolyglotBridgeCall handler to enforce proof obligations
 * that Betty Phase 8 identified at compile time.
 */

import type { TopologyType, JsonSchema } from './semantic-compatibility';

export interface ValidationResult {
  valid: boolean;
  errors: ValidationError[];
}

export interface ValidationError {
  path: string;
  expected: string;
  actual: string;
  message: string;
}

/**
 * Validate a JSON value against a TopologyType contract.
 */
export function validateValue(
  value: unknown,
  expectedType: TopologyType,
  path: string = '$'
): ValidationResult {
  const errors: ValidationError[] = [];
  validateRecursive(value, expectedType, path, errors);
  return { valid: errors.length === 0, errors };
}

function validateRecursive(
  value: unknown,
  expectedType: TopologyType,
  path: string,
  errors: ValidationError[]
): void {
  switch (expectedType.kind) {
    case 'unknown':
      // No constraint -- always valid.
      return;

    case 'json':
      validateJsonSchema(value, expectedType.schema, path, errors);
      return;

    case 'bytes':
      // At the JSON boundary, bytes should be a base64 string or Uint8Array.
      if (
        typeof value !== 'string' &&
        !(value instanceof Uint8Array) &&
        !ArrayBuffer.isView(value)
      ) {
        errors.push({
          path,
          expected: 'bytes (string | Uint8Array)',
          actual: typeof value,
          message: `Expected bytes at ${path}, got ${typeof value}`,
        });
      }
      return;

    case 'stream':
      if (!Array.isArray(value)) {
        errors.push({
          path,
          expected: `Stream<${formatTypeShort(expectedType.element)}>`,
          actual: typeof value,
          message: `Expected array at ${path}, got ${typeof value}`,
        });
        return;
      }
      for (let i = 0; i < value.length; i++) {
        validateRecursive(
          value[i],
          expectedType.element,
          `${path}[${i}]`,
          errors
        );
      }
      return;

    case 'option':
      if (value === null || value === undefined) {
        // null/undefined is valid for Option.
        return;
      }
      validateRecursive(value, expectedType.inner, path, errors);
      return;

    case 'product':
      if (typeof value !== 'object' || value === null || Array.isArray(value)) {
        errors.push({
          path,
          expected: 'object (product type)',
          actual: Array.isArray(value) ? 'array' : typeof value,
          message: `Expected object at ${path}, got ${
            Array.isArray(value) ? 'array' : typeof value
          }`,
        });
        return;
      }
      // Check required fields (closed product).
      if (!expectedType.open) {
        for (const [fieldName, fieldType] of expectedType.fields) {
          if (!(fieldName in (value as Record<string, unknown>))) {
            errors.push({
              path: `${path}.${fieldName}`,
              expected: formatTypeShort(fieldType),
              actual: 'missing',
              message: `Required field '${fieldName}' missing at ${path}`,
            });
          } else {
            validateRecursive(
              (value as Record<string, unknown>)[fieldName],
              fieldType,
              `${path}.${fieldName}`,
              errors
            );
          }
        }
      }
      return;

    case 'sum':
      // Sum types are harder to validate without a discriminator.
      // Check if the value matches any variant.
      if (typeof value === 'object' && value !== null) {
        // Look for a type/kind/tag discriminator.
        const obj = value as Record<string, unknown>;
        const tag = obj['type'] ?? obj['kind'] ?? obj['tag'] ?? obj['variant'];
        if (typeof tag === 'string') {
          const variant = expectedType.variants.find(([name]) => name === tag);
          if (variant) {
            validateRecursive(
              obj['value'] ?? obj['data'] ?? obj,
              variant[1],
              path,
              errors
            );
            return;
          }
        }
      }
      // Can't validate sum type without discriminator -- pass.
      return;

    case 'opaque':
      // Opaque types can't be validated at the JSON boundary.
      // This is always a proof obligation.
      return;
  }
}

function validateJsonSchema(
  value: unknown,
  schema: JsonSchema,
  path: string,
  errors: ValidationError[]
): void {
  switch (schema.type) {
    case 'any':
      return;

    case 'null':
      if (value !== null && value !== undefined) {
        errors.push({
          path,
          expected: 'null',
          actual: typeof value,
          message: `Expected null at ${path}, got ${typeof value}`,
        });
      }
      return;

    case 'boolean':
      if (typeof value !== 'boolean') {
        errors.push({
          path,
          expected: 'boolean',
          actual: typeof value,
          message: `Expected boolean at ${path}, got ${typeof value}`,
        });
      }
      return;

    case 'number':
      if (typeof value !== 'number') {
        errors.push({
          path,
          expected: 'number',
          actual: typeof value,
          message: `Expected number at ${path}, got ${typeof value}`,
        });
      }
      return;

    case 'integer':
      if (typeof value !== 'number') {
        errors.push({
          path,
          expected: 'integer',
          actual: typeof value,
          message: `Expected integer at ${path}, got ${typeof value}`,
        });
      } else if (!Number.isInteger(value)) {
        errors.push({
          path,
          expected: 'integer',
          actual: `number (${value})`,
          message: `Expected integer at ${path}, got non-integer number ${value}`,
        });
      }
      return;

    case 'string':
      if (typeof value !== 'string') {
        errors.push({
          path,
          expected: 'string',
          actual: typeof value,
          message: `Expected string at ${path}, got ${typeof value}`,
        });
      }
      return;

    case 'array':
      if (!Array.isArray(value)) {
        errors.push({
          path,
          expected: 'array',
          actual: typeof value,
          message: `Expected array at ${path}, got ${typeof value}`,
        });
        return;
      }
      for (let i = 0; i < value.length; i++) {
        validateJsonSchema(value[i], schema.items, `${path}[${i}]`, errors);
      }
      return;

    case 'object':
      if (typeof value !== 'object' || value === null || Array.isArray(value)) {
        errors.push({
          path,
          expected: 'object',
          actual: Array.isArray(value) ? 'array' : typeof value,
          message: `Expected object at ${path}, got ${
            Array.isArray(value) ? 'array' : typeof value
          }`,
        });
        return;
      }
      for (const [propName, propSchema] of schema.properties) {
        if (propName in (value as Record<string, unknown>)) {
          validateJsonSchema(
            (value as Record<string, unknown>)[propName],
            propSchema,
            `${path}.${propName}`,
            errors
          );
        } else if (!schema.additional) {
          errors.push({
            path: `${path}.${propName}`,
            expected: formatSchemaShort(propSchema),
            actual: 'missing',
            message: `Required property '${propName}' missing at ${path}`,
          });
        }
      }
      return;
  }
}

function formatTypeShort(ty: TopologyType): string {
  switch (ty.kind) {
    case 'json':
      return ty.schema.type;
    case 'bytes':
      return 'bytes';
    case 'stream':
      return `[${formatTypeShort(ty.element)}]`;
    case 'option':
      return `${formatTypeShort(ty.inner)}?`;
    case 'product':
      return 'object';
    case 'sum':
      return ty.variants.map(([n]) => n).join('|');
    case 'opaque':
      return `${ty.language}::${ty.name}`;
    case 'unknown':
      return 'any';
  }
}

function formatSchemaShort(schema: JsonSchema): string {
  return schema.type;
}
