// @a0n/.ignored_gnosis/src/betty/semantic-healer.ts
/**
 * Betty Semantic Healer -- auto-insert type converter nodes at incompatible edges.
 *
 * When Betty Phase 8 detects that a cross-language edge has incompatible types
 * or a proof obligation, the healer inserts a converter node that transforms
 * the data between the source and target types.
 *
 * The healing is itself a fork/race/fold: fork candidate converters, race them
 * against the type system, fold to the one that resolves the obligation.
 */

import type { GraphAST } from './compiler';
import type {
  TopologyType,
  SemanticCompatibilityResult,
  EdgeSemanticCheck,
} from './semantic-compatibility';
import {
  checkTypeCompatibility,
  formatTopologyType,
} from './semantic-compatibility';

// ─── Converter Registry ─────────────────────────────────────────────────────

/**
 * A type converter that can be inserted into a topology to bridge
 * incompatible types.
 */
export interface TypeConverter {
  /** Unique converter ID. */
  id: string;
  /** Human-readable name. */
  name: string;
  /** Source type this converter accepts. */
  fromType: TopologyType;
  /** Target type this converter produces. */
  toType: TopologyType;
  /** The conversion strategy (used as a GG node property). */
  strategy: string;
  /** Description of what the converter does. */
  description: string;
  /** Whether this converter is lossless (invertible). */
  lossless: boolean;
  /** Priority (lower = preferred). */
  priority: number;
}

/**
 * Built-in converter registry. Each converter bridges a specific
 * source→target type pair.
 */
export const CONVERTER_REGISTRY: TypeConverter[] = [
  // ─── Bytes ↔ String ────────────────────────────────────────────────
  {
    id: 'bytes-to-base64',
    name: 'Base64 Encode',
    fromType: { kind: 'bytes' },
    toType: { kind: 'json', schema: { type: 'string' } },
    strategy: 'base64_encode',
    description: 'Encode bytes as base64 string for JSON transport',
    lossless: true,
    priority: 1,
  },
  {
    id: 'base64-to-bytes',
    name: 'Base64 Decode',
    fromType: { kind: 'json', schema: { type: 'string' } },
    toType: { kind: 'bytes' },
    strategy: 'base64_decode',
    description: 'Decode base64 string back to bytes',
    lossless: true,
    priority: 1,
  },
  {
    id: 'bytes-to-hex',
    name: 'Hex Encode',
    fromType: { kind: 'bytes' },
    toType: { kind: 'json', schema: { type: 'string' } },
    strategy: 'hex_encode',
    description: 'Encode bytes as hex string',
    lossless: true,
    priority: 2,
  },
  // ─── Number ↔ Integer ──────────────────────────────────────────────
  {
    id: 'number-to-integer-floor',
    name: 'Floor to Integer',
    fromType: { kind: 'json', schema: { type: 'number' } },
    toType: { kind: 'json', schema: { type: 'integer' } },
    strategy: 'math_floor',
    description: 'Convert number to integer via Math.floor()',
    lossless: false,
    priority: 1,
  },
  {
    id: 'number-to-integer-round',
    name: 'Round to Integer',
    fromType: { kind: 'json', schema: { type: 'number' } },
    toType: { kind: 'json', schema: { type: 'integer' } },
    strategy: 'math_round',
    description: 'Convert number to integer via Math.round()',
    lossless: false,
    priority: 2,
  },
  {
    id: 'number-to-integer-check',
    name: 'Assert Integer',
    fromType: { kind: 'json', schema: { type: 'number' } },
    toType: { kind: 'json', schema: { type: 'integer' } },
    strategy: 'assert_integer',
    description: 'Assert value is already an integer, throw if not',
    lossless: true,
    priority: 0, // Preferred: no data loss, fails fast.
  },
  // ─── Product ↔ JSON ────────────────────────────────────────────────
  {
    id: 'product-to-json-object',
    name: 'Product to JSON Object',
    fromType: { kind: 'product', fields: [], open: true },
    toType: {
      kind: 'json',
      schema: { type: 'object', properties: [], additional: true },
    },
    strategy: 'json_serialize',
    description: 'Serialize product type to JSON object',
    lossless: true,
    priority: 1,
  },
  {
    id: 'json-object-to-product',
    name: 'JSON Object to Product',
    fromType: {
      kind: 'json',
      schema: { type: 'object', properties: [], additional: true },
    },
    toType: { kind: 'product', fields: [], open: true },
    strategy: 'json_deserialize',
    description: 'Deserialize JSON object to product type',
    lossless: true,
    priority: 1,
  },
  // ─── Stream ↔ JSON Array ───────────────────────────────────────────
  {
    id: 'stream-to-json-array',
    name: 'Stream to JSON Array',
    fromType: {
      kind: 'stream',
      element: { kind: 'json', schema: { type: 'any' } },
    },
    toType: { kind: 'json', schema: { type: 'array', items: { type: 'any' } } },
    strategy: 'collect_to_array',
    description: 'Collect stream elements into JSON array',
    lossless: true,
    priority: 1,
  },
  // ─── Option unwrap ─────────────────────────────────────────────────
  {
    id: 'option-unwrap',
    name: 'Unwrap Option',
    fromType: {
      kind: 'option',
      inner: { kind: 'json', schema: { type: 'any' } },
    },
    toType: { kind: 'json', schema: { type: 'any' } },
    strategy: 'unwrap_or_throw',
    description: 'Unwrap optional value, throw if null/None/nil',
    lossless: false,
    priority: 1,
  },
  {
    id: 'option-unwrap-default',
    name: 'Unwrap with Default',
    fromType: {
      kind: 'option',
      inner: { kind: 'json', schema: { type: 'any' } },
    },
    toType: { kind: 'json', schema: { type: 'any' } },
    strategy: 'unwrap_or_default',
    description: 'Unwrap optional value, use default if null/None/nil',
    lossless: false,
    priority: 2,
  },
  // ─── Sum → Product (Result → value) ───────────────────────────────
  {
    id: 'result-unwrap',
    name: 'Unwrap Result',
    fromType: {
      kind: 'sum',
      variants: [
        ['Ok', { kind: 'json', schema: { type: 'any' } }],
        ['Err', { kind: 'json', schema: { type: 'any' } }],
      ],
    },
    toType: { kind: 'json', schema: { type: 'any' } },
    strategy: 'unwrap_result',
    description: 'Unwrap Result/Either, throw on Err',
    lossless: false,
    priority: 1,
  },
  // ─── String ↔ JSON parse ───────────────────────────────────────────
  {
    id: 'string-to-json',
    name: 'JSON Parse',
    fromType: { kind: 'json', schema: { type: 'string' } },
    toType: { kind: 'json', schema: { type: 'any' } },
    strategy: 'json_parse',
    description: 'Parse JSON string into structured value',
    lossless: true,
    priority: 1,
  },
  {
    id: 'json-to-string',
    name: 'JSON Stringify',
    fromType: { kind: 'json', schema: { type: 'any' } },
    toType: { kind: 'json', schema: { type: 'string' } },
    strategy: 'json_stringify',
    description: 'Serialize value to JSON string',
    lossless: true,
    priority: 1,
  },
];

// ─── Converter Lookup ───────────────────────────────────────────────────────

/**
 * Find converters that can bridge a source→target type gap.
 * Returns candidates sorted by priority (lower = better).
 */
export function findConverters(
  sourceType: TopologyType,
  targetType: TopologyType
): TypeConverter[] {
  const candidates: TypeConverter[] = [];

  for (const converter of CONVERTER_REGISTRY) {
    // Check if the converter's fromType is compatible with sourceType
    // and its toType is compatible with targetType.
    const fromCompat = checkTypeCompatibility(sourceType, converter.fromType);
    const toCompat = checkTypeCompatibility(converter.toType, targetType);

    if (
      (fromCompat.status === 'compatible' ||
        fromCompat.status === 'proof-obligation') &&
      (toCompat.status === 'compatible' ||
        toCompat.status === 'proof-obligation')
    ) {
      candidates.push(converter);
    }
  }

  // Also check structural matches for parameterized types.
  // E.g., Stream(Json(number)) → Bytes needs bytes-to-base64 in reverse.
  if (sourceType.kind === 'bytes' && targetType.kind !== 'bytes') {
    const bytesConverters = CONVERTER_REGISTRY.filter(
      (c) => c.fromType.kind === 'bytes'
    );
    for (const c of bytesConverters) {
      if (!candidates.includes(c)) {
        const toCompat = checkTypeCompatibility(c.toType, targetType);
        if (
          toCompat.status === 'compatible' ||
          toCompat.status === 'proof-obligation'
        ) {
          candidates.push(c);
        }
      }
    }
  }

  return candidates.sort((a, b) => a.priority - b.priority);
}

// ─── Healing Result ─────────────────────────────────────────────────────────

export interface HealingResult {
  /** The healed AST with converter nodes inserted. */
  ast: GraphAST;
  /** Report of insertions. */
  insertions: ConverterInsertion[];
  /** Edges that could not be healed (no converter found). */
  unhealable: UnhealableEdge[];
  /** Total number of converters inserted. */
  converterCount: number;
}

export interface ConverterInsertion {
  /** The original edge source. */
  sourceNode: string;
  /** The original edge target. */
  targetNode: string;
  /** The converter node ID inserted. */
  converterNodeId: string;
  /** The converter used. */
  converter: TypeConverter;
  /** Whether this healed an incompatibility or resolved a proof obligation. */
  healedStatus: 'incompatible' | 'proof-obligation';
}

export interface UnhealableEdge {
  sourceNode: string;
  targetNode: string;
  sourceType: TopologyType;
  targetType: TopologyType;
  reason: string;
}

// ─── Healer ─────────────────────────────────────────────────────────────────

/**
 * Heal a topology by inserting converter nodes at incompatible edges.
 *
 * For each edge flagged by Phase 8:
 * 1. Fork: find all candidate converters from the registry.
 * 2. Race: check which converter produces a type compatible with the target.
 * 3. Fold: select the highest-priority converter.
 * 4. Insert: splice the converter node into the edge.
 */
export function healTopology(
  ast: GraphAST,
  semanticResult: SemanticCompatibilityResult
): HealingResult {
  // Clone the AST so we don't mutate the original.
  const healedNodes = new Map(ast.nodes);
  let healedEdges = [...ast.edges];
  const insertions: ConverterInsertion[] = [];
  const unhealable: UnhealableEdge[] = [];

  // Collect edges that need healing.
  const edgesToHeal: EdgeSemanticCheck[] = semanticResult.edgeChecks.filter(
    (check) =>
      check.compatibility.status === 'incompatible' ||
      check.compatibility.status === 'proof-obligation'
  );

  for (const check of edgesToHeal) {
    // Fork: find candidate converters.
    const candidates = findConverters(check.sourceType, check.targetType);

    if (candidates.length === 0) {
      unhealable.push({
        sourceNode: check.sourceNode,
        targetNode: check.targetNode,
        sourceType: check.sourceType,
        targetType: check.targetType,
        reason: `no converter found for ${formatTopologyType(
          check.sourceType
        )} → ${formatTopologyType(check.targetType)}`,
      });
      continue;
    }

    // Race/Fold: pick the best converter (lowest priority number).
    const winner = candidates[0]!;

    // Insert: create a converter node and splice it into the edge.
    const converterNodeId = `auto_${winner.id}_${check.sourceNode}_${check.targetNode}`;

    // Create the converter node.
    healedNodes.set(converterNodeId, {
      id: converterNodeId,
      labels: ['Statement', 'TypeConverter'],
      properties: {
        converter_id: winner.id,
        converter_name: winner.name,
        strategy: winner.strategy,
        from_type: formatTopologyType(check.sourceType),
        to_type: formatTopologyType(check.targetType),
        lossless: String(winner.lossless),
        description: winner.description,
        auto_inserted: 'true',
      },
    });

    // Splice: replace source→target with source→converter→target.
    healedEdges = healedEdges.map((edge) => {
      const sourceIdx = edge.sourceIds.indexOf(check.sourceNode);
      const targetIdx = edge.targetIds.indexOf(check.targetNode);

      if (sourceIdx >= 0 && targetIdx >= 0) {
        // This edge connects source to target. Replace target with converter.
        const newTargetIds = [...edge.targetIds];
        newTargetIds[targetIdx] = converterNodeId;
        return { ...edge, targetIds: newTargetIds };
      }
      return edge;
    });

    // Add converter→target edge.
    healedEdges.push({
      sourceIds: [converterNodeId],
      targetIds: [check.targetNode],
      type: 'PROCESS',
      properties: { auto_healed: 'true' },
    });

    insertions.push({
      sourceNode: check.sourceNode,
      targetNode: check.targetNode,
      converterNodeId,
      converter: winner,
      healedStatus: check.compatibility.status as
        | 'incompatible'
        | 'proof-obligation',
    });
  }

  return {
    ast: { nodes: healedNodes, edges: healedEdges },
    insertions,
    unhealable,
    converterCount: insertions.length,
  };
}

/**
 * Format a healing result as a human-readable report.
 */
export function formatHealingReport(result: HealingResult): string {
  const lines: string[] = [];

  if (result.insertions.length > 0) {
    lines.push(`Healed ${result.converterCount} edge(s):`);
    for (const ins of result.insertions) {
      const status = ins.healedStatus === 'incompatible' ? 'x->[ok]' : '[warn]->[ok]';
      lines.push(
        `  ${status} ${ins.sourceNode} → ${ins.targetNode}: inserted ${ins.converter.name} (${ins.converter.strategy})`
      );
      if (!ins.converter.lossless) {
        lines.push(`      [warn] lossy conversion: ${ins.converter.description}`);
      }
    }
  }

  if (result.unhealable.length > 0) {
    lines.push(`${result.unhealable.length} unhealable edge(s):`);
    for (const uh of result.unhealable) {
      lines.push(`  ✗ ${uh.sourceNode} → ${uh.targetNode}: ${uh.reason}`);
    }
  }

  if (result.insertions.length === 0 && result.unhealable.length === 0) {
    lines.push('No healing needed — all edges are type-compatible.');
  }

  return lines.join('\n');
}
