// @a0n/.ignored_gnosis/src/betty/teleportation.ts
/**
 * Code generation for teleportation edges.
 *
 * When the compiler sees TUNNEL with teleport: 'true', generate code
 * that sends only the deficit integer instead of full data. The deficit
 * alone is a sufficient statistic for determining trajectory.
 */

import type { ASTEdge, Diagnostic } from './compiler';

export interface TeleportationEdge {
  edge: ASTEdge;
  deficitOnly: boolean;
}

/**
 * Analyze TUNNEL edges for teleportation optimization.
 * Returns edges that can be optimized to send only deficit.
 */
export function analyzeTeleportationEdges(
  edges: ASTEdge[]
): TeleportationEdge[] {
  return edges
    .filter((e) => e.type === 'TUNNEL')
    .map((edge) => ({
      edge,
      deficitOnly: edge.properties.teleport === 'true',
    }));
}

/**
 * Validate teleportation edges.
 */
export function validateTeleportation(edges: ASTEdge[]): Diagnostic[] {
  const diagnostics: Diagnostic[] = [];
  const teleportEdges = analyzeTeleportationEdges(edges);

  for (const { edge, deficitOnly } of teleportEdges) {
    if (deficitOnly && edge.sourceIds.length > 1) {
      diagnostics.push({
        line: 1,
        column: 1,
        message: `TUNNEL with teleport: 'true' from [${edge.sourceIds.join(
          ', '
        )}] has multiple sources. Teleportation sends a single deficit integer -- use a FOLD first to merge sources.`,
        severity: 'warning',
      });
    }
  }

  return diagnostics;
}
