// @a0n/.ignored_gnosis/src/betty/ultrametric-allocator.ts
/**
 * The Ultrametric Memory Allocator
 * 
 * Replaces linear Archimedean array allocation with p-Adic sphere allocation.
 * Completely eliminates memory fragmentation and buffer overflows by enforcing
 * the geometric impossibility of partial intersection.
 */

import { UltrametricSphere, PRIME_BASE, pAdicNorm } from './p-adic-topology';

export interface MemoryPayload {
  data: string;
}

export class UltrametricAllocator {
  private spheres: Map<number, { sphere: UltrametricSphere, payload: MemoryPayload }> = new Map();
  private p: number;

  constructor(p: number = PRIME_BASE) {
    this.p = p;
  }

  /**
   * Allocates a new p-Adic memory boundary.
   * 
   * @param center The topological center coordinate
   * @param radius The p-adic radius (expansion bound)
   * @param payload The data payload
   */
  public allocate(center: number, radius: number, payload: MemoryPayload): void {
    const newSphere = new UltrametricSphere(center, radius, this.p);

    // Verify geometric isolation
    for (const [existingCenter, allocation] of this.spheres.entries()) {
      const state = newSphere.getIntersectionState(allocation.sphere);
      
      if (state === 'IDENTICAL') {
        throw new Error(`Allocation Collision: Exact sphere already exists at center ${existingCenter}`);
      }
      if (state === 'ENCAPSULATED_BY') {
        // Hierarchical nesting is allowed.
        console.log(`[Allocator] Memory successfully nested within existing sphere ${existingCenter}`);
      }
      if (state === 'ENCAPSULATES') {
        // Re-allocation expansion is allowed.
        console.log(`[Allocator] Memory allocation encompasses existing sphere ${existingCenter}`);
      }
      if (state === 'DISJOINT') {
        // Perfect isolation.
      }
      
      // Note: 'PARTIAL_INTERSECTION' is impossible to return, proving that 
      // Archimedean buffer bleeds cannot physically occur in this topology.
    }

    this.spheres.set(center, { sphere: newSphere, payload });
  }

  /**
   * Reads from an exact linear pointer, routing it through the p-adic norm 
   * to guarantee it resolves strictly within isolated boundaries.
   */
  public read(pointer: number): MemoryPayload | null {
    // A standard linear buffer overflow involves pointer++ or pointer + offset.
    // In p-adic space, stepping +1 linearly shifts the topological distance by |1|_p = 1.
    // If the allocated sphere radius is < 1, the step instantly vaults you out of the sphere 
    // into the void, rather than bleeding into an adjacent array index.
    
    for (const [center, allocation] of this.spheres.entries()) {
      if (allocation.sphere.contains(pointer)) {
        return allocation.payload;
      }
    }
    
    return null; // Segmentation Fault: Void Access
  }

  public getSphereCount(): number {
    return this.spheres.size;
  }
}
