// @a0n/.ignored_gnosis/src/universe/index.ts
export {
  GOD_FORMULA_KERNEL,
  evaluateLawProgram,
  verify,
  verifyBuleyean,
} from './clockwork';
export type { ClockworkConfig } from './clockwork';

export type {
  ClockworkVerdict,
  VerdictStatus,
  VerdictCheck,
  ConservationLedger,
  DeficitSummary,
  VoidSnapshot,
  HeatLedger,
  LaplaceWitness,
} from './verdict';
export {
  passCheck,
  failCheck,
  verdictStatus,
  verdictSummary,
} from './verdict';

export {
  generateCertificate,
  generateUniverseCertificate,
  attachUniverseRuntimeMetadata,
  serializeCertificate,
  serializeUniverseCertificate,
  deserializeCertificate,
  deserializeUniverseCertificate,
  auditCertificate,
  auditUniverseCertificate,
} from './certificate';
export type { ClockworkCertificate } from './certificate';

export {
  LAW_KERNEL_ID,
  LAW_DIMENSIONS,
  APPLICATION_DIMENSION,
  CURRENT_DIMENSION,
  UNIVERSE_LANGUAGE_CODES,
  QUANTUM_BASES,
  QUANTUM_LITERALS,
  UNIVERSE_CONSTANTS,
  UNIVERSE_FORMULAS,
  createQuantumLiteral,
  createLawState,
  computeLawProgramMetrics,
  canonicalizeLawProgram,
  digestLawProgram,
  sha256Hex,
  stableStringify,
} from './kernel';
export type {
  LawKernelId,
  LawDimension,
  UniverseLanguageCode,
  QuantumBasis,
  QuantumLiteral,
  QuantumRegisterLiteral,
  LawNode,
  LawEdge,
  LawCapabilitySummary,
  LawProgram,
  LawProgramMetrics,
  UniverseBuildMetadata,
  UniverseRuntimeMetadata,
  UniverseCertificate,
  UniverseAdmissionReport,
  LawState,
  LawKernel,
  LawKernelEvaluation,
} from './kernel';

export {
  UNIVERSAL_EPOCH_REFERENCE_AGE_YEARS,
  SOLAR_YEAR_SECONDS,
  UNIVERSAL_EPOCH_SECONDS_AT_UNIX_EPOCH,
  quasicrystallineVectorClock,
  secondsSinceUniversalEpoch,
  gnosisUniversalEpochClock,
} from './universal-epoch-clock';
export type {
  GnosisQuasicrystallineVectorClock,
  GnosisUniversalEpochClock,
} from './universal-epoch-clock';

export {
  checkSevenLaws,
  checkGodFormula,
  checkControlStatistics,
  checkReductionChain,
  fullBuleyeanAudit,
  auditVoidBoundary,
  formatAuditReport,
} from './audit';
export type {
  CheckResult,
  SevenLawsReport,
  GodFormulaReport,
  ControlStatisticReport,
  ReductionChainReport,
  FullAuditReport,
  BuleyeanInput,
} from './audit';
