// @a0n/.ignored_gnosis/src/universe/kernel.ts
import type { ClockworkCertificate } from './certificate';
import type { ClockworkVerdict, VerdictStatus } from './verdict';

export const LAW_KERNEL_ID = 'god-formula/v1' as const;

export type LawKernelId = typeof LAW_KERNEL_ID;

export const LAW_DIMENSIONS = [
  'dimensionless',
  'physical',
  'information',
  'language',
  'communication',
  'quantum',
  'temporal',
  'distance',
] as const;

export type LawDimension = (typeof LAW_DIMENSIONS)[number];

export const CURRENT_DIMENSION = 'information' as const;
export const APPLICATION_DIMENSION = 'dimensionless' as const;

export const UNIVERSE_LANGUAGE_CODES = [
  'en',
  'es',
  'zh',
  'ar',
  'fr',
  'de',
] as const;

export type UniverseLanguageCode = (typeof UNIVERSE_LANGUAGE_CODES)[number];

export const QUANTUM_BASES = ['0', '1', '+', '-'] as const;
export type QuantumBasis = (typeof QUANTUM_BASES)[number];

export interface QuantumLiteral {
  readonly kind: 'qubit';
  readonly basis: QuantumBasis;
  readonly alpha: number;
  readonly beta: number;
  readonly dimension: typeof APPLICATION_DIMENSION;
  readonly currentDimension: 'quantum';
}

export interface QuantumRegisterLiteral {
  readonly kind: 'register';
  readonly name: string;
  readonly qubits: readonly QuantumLiteral[];
  readonly entangled: boolean;
  readonly dimension: typeof APPLICATION_DIMENSION;
  readonly currentDimension: 'quantum';
}

export interface LawNode {
  readonly id: string;
  readonly labels: readonly string[];
  readonly properties: Readonly<Record<string, string>>;
}

export interface LawEdge {
  readonly sourceIds: readonly string[];
  readonly targetIds: readonly string[];
  readonly type: string;
  readonly properties: Readonly<Record<string, string>>;
}

export interface LawCapabilitySummary {
  readonly target?: string;
  readonly required?: readonly string[];
  readonly declared?: readonly string[];
  readonly inferred?: readonly string[];
  readonly importedSurface?: readonly string[];
  readonly issueCount?: number;
}

export interface LawProgram {
  readonly nodes: ReadonlyMap<string, LawNode>;
  readonly edges: readonly LawEdge[];
  readonly subject?: string;
  readonly sourceDigest?: string;
  readonly capabilitySummary?: LawCapabilitySummary;
  readonly dependencyCertificateDigests?: readonly string[];
  readonly metadata?: Readonly<Record<string, string | number | boolean>>;
}

export interface LawProgramMetrics {
  readonly nodeCount: number;
  readonly edgeCount: number;
  readonly dimensionCount: number;
}

export interface UniverseBuildMetadata {
  readonly verifiedAt: number;
  readonly nodeCount: number;
  readonly edgeCount: number;
  readonly dimensionCount: number;
  readonly applicationDimension: typeof APPLICATION_DIMENSION;
  readonly currentDimension: typeof CURRENT_DIMENSION;
  readonly dimensionlessApplications: true;
  readonly quantumPrimitivesExplicit: true;
}

export interface UniverseRuntimeMetadata {
  readonly executedAt: number;
  readonly entrypoint?: string;
  readonly runtimeName?: string;
}

export interface UniverseCertificate {
  readonly version: 1;
  readonly kernelId: LawKernelId;
  readonly subject: string;
  readonly status: VerdictStatus;
  readonly theoremRefs: readonly string[];
  readonly capabilitySummary: LawCapabilitySummary;
  readonly dependencyCertificateDigests: readonly string[];
  readonly sourceDigest: string;
  readonly buildMetadata: UniverseBuildMetadata;
  readonly runtimeMetadata?: UniverseRuntimeMetadata;
  readonly dimension: typeof APPLICATION_DIMENSION;
  readonly currentDimension: typeof CURRENT_DIMENSION;
  readonly clockwork: ClockworkCertificate;
  readonly digest: string;
}

export interface UniverseAdmissionReport {
  readonly kernelId: LawKernelId;
  readonly status: VerdictStatus;
  readonly ok: boolean;
  readonly correctnessOk: boolean;
  readonly capabilityOk: boolean;
  readonly reasons: readonly string[];
  readonly certificate: UniverseCertificate;
  readonly law: LawState;
}

export interface LawState {
  readonly kernelId: LawKernelId;
  readonly status: VerdictStatus;
  readonly dimension: typeof APPLICATION_DIMENSION;
  readonly currentDimension: typeof CURRENT_DIMENSION;
  readonly deficit: number;
  readonly heat: number;
  readonly conservation: boolean;
  readonly certificateDigest: string;
  readonly theoremRefs: readonly string[];
  readonly certificate: UniverseCertificate;
  readonly constants: typeof UNIVERSE_CONSTANTS;
  readonly formulas: typeof UNIVERSE_FORMULAS;
  readonly quantum: typeof QUANTUM_LITERALS;
}

export interface LawKernelEvaluation {
  readonly verdict: ClockworkVerdict;
  readonly clockworkCertificate: ClockworkCertificate;
  readonly certificate: UniverseCertificate;
  readonly law: LawState;
}

export interface LawKernel {
  readonly id: LawKernelId;
  verify(program: LawProgram): LawKernelEvaluation;
  readonly constants: typeof UNIVERSE_CONSTANTS;
  readonly formulas: typeof UNIVERSE_FORMULAS;
}

/**
 * Creates the Quantum Literal.
 */
export function createQuantumLiteral(basis: QuantumBasis): QuantumLiteral {
  switch (basis) {
    case '1':
      return {
        kind: 'qubit',
        basis,
        alpha: 0,
        beta: 1,
        dimension: APPLICATION_DIMENSION,
        currentDimension: 'quantum',
      };
    case '+':
      return {
        kind: 'qubit',
        basis,
        alpha: Math.SQRT1_2,
        beta: Math.SQRT1_2,
        dimension: APPLICATION_DIMENSION,
        currentDimension: 'quantum',
      };
    case '-':
      return {
        kind: 'qubit',
        basis,
        alpha: Math.SQRT1_2,
        beta: -Math.SQRT1_2,
        dimension: APPLICATION_DIMENSION,
        currentDimension: 'quantum',
      };
    case '0':
    default:
      return {
        kind: 'qubit',
        basis: '0',
        alpha: 1,
        beta: 0,
        dimension: APPLICATION_DIMENSION,
        currentDimension: 'quantum',
      };
  }
}

export const QUANTUM_LITERALS = {
  ZERO: createQuantumLiteral('0'),
  ONE: createQuantumLiteral('1'),
  PLUS: createQuantumLiteral('+'),
  MINUS: createQuantumLiteral('-'),
  BELL: {
    kind: 'register',
    name: 'bell',
    qubits: [createQuantumLiteral('+'), createQuantumLiteral('0')],
    entangled: true,
    dimension: APPLICATION_DIMENSION,
    currentDimension: 'quantum',
  } satisfies QuantumRegisterLiteral,
} as const;

export const UNIVERSE_CONSTANTS = {
  math: {
    PI: Math.PI,
    TAU: Math.PI * 2,
    E: Math.E,
    SQRT2: Math.SQRT2,
    SQRT1_2: Math.SQRT1_2,
    LN2: Math.LN2,
  },
  physics: {
    SPEED_OF_LIGHT_MPS: 299_792_458,
    PLANCK_CONSTANT_J_S: 6.626_070_15e-34,
    REDUCED_PLANCK_CONSTANT_J_S: 1.054_571_817e-34,
    GRAVITATIONAL_CONSTANT: 6.674_30e-11,
    BOLTZMANN_CONSTANT: 1.380_649e-23,
    ELEMENTARY_CHARGE_C: 1.602_176_634e-19,
    AVOGADRO_CONSTANT: 6.022_140_76e23,
    VACUUM_PERMITTIVITY: 8.854_187_812_8e-12,
    VACUUM_PERMEABILITY: 1.256_637_062_12e-6,
  },
  time: {
    SECOND: 1,
    MILLISECOND: 1e-3,
    MICROSECOND: 1e-6,
    MINUTE: 60,
    HOUR: 3_600,
    DAY: 86_400,
    PICOLORENZO_SECONDS: Math.PI * 86_400,
    AEON_SECONDS: 110 * Math.PI * 86_400,
  },
  distance: {
    METER: 1,
    KILOMETER: 1_000,
    CENTIMETER: 1e-2,
    MILLIMETER: 1e-3,
    ASTRONOMICAL_UNIT_METERS: 149_597_870_700,
    LIGHT_YEAR_METERS: 9.460_730_472_580_8e15,
    PARSEC_METERS: 3.085_677_581_491_367e16,
  },
  dimension: {
    APPLICATION: APPLICATION_DIMENSION,
    CURRENT: CURRENT_DIMENSION,
    ALL: LAW_DIMENSIONS,
    PHYSICAL: 'physical',
    INFORMATION: 'information',
    LANGUAGE: 'language',
    COMMUNICATION: 'communication',
    QUANTUM: 'quantum',
    TEMPORAL: 'temporal',
    DISTANCE: 'distance',
  },
  language: {
    DEFAULT: 'en',
    SUPPORTED: UNIVERSE_LANGUAGE_CODES,
    STREAMS: 1,
    DISCOURSE_CHANNELS: 1,
  },
  communication: {
    THOUGHT_DIMENSIONS_FLOOR: 1,
    OUTPUT_STREAMS: 1,
    SEMIOTIC_CHANNELS: 1,
  },
  quantum: {
    BASIS: QUANTUM_BASES,
    LITERALS: QUANTUM_LITERALS,
  },
} as const;

function spatialDistanceSquared(
  dxMeters: number,
  dyMeters: number = 0,
  dzMeters: number = 0
): number {
  return dxMeters ** 2 + dyMeters ** 2 + dzMeters ** 2;
}

export const UNIVERSE_FORMULAS = {
  godFormulaWeight(rounds: number, rejections: number): number {
    return rounds - Math.min(rejections, rounds) + 1;
  },
  complementWeight(rounds: number, rejections: number): number {
    return rounds - Math.min(rejections, rounds) + 1;
  },
  landauerHeat(bitsErased: number, temperatureKelvin: number = 1): number {
    return (
      UNIVERSE_CONSTANTS.physics.BOLTZMANN_CONSTANT *
      temperatureKelvin *
      Math.LN2 *
      bitsErased
    );
  },
  energyMassEquivalence(massKg: number): number {
    return massKg * UNIVERSE_CONSTANTS.physics.SPEED_OF_LIGHT_MPS ** 2;
  },
  photonEnergy(frequencyHz: number): number {
    return UNIVERSE_CONSTANTS.physics.PLANCK_CONSTANT_J_S * frequencyHz;
  },
  gravitationalForce(
    massAKg: number,
    massBKg: number,
    distanceMeters: number
  ): number {
    if (distanceMeters === 0) {
      return Number.POSITIVE_INFINITY;
    }

    return (
      (UNIVERSE_CONSTANTS.physics.GRAVITATIONAL_CONSTANT * massAKg * massBKg) /
      distanceMeters ** 2
    );
  },
  schwarzschildRadius(massKg: number): number {
    return (
      (2 * UNIVERSE_CONSTANTS.physics.GRAVITATIONAL_CONSTANT * massKg) /
      UNIVERSE_CONSTANTS.physics.SPEED_OF_LIGHT_MPS ** 2
    );
  },
  spacetimeIntervalSquared(
    deltaSeconds: number,
    dxMeters: number,
    dyMeters: number = 0,
    dzMeters: number = 0
  ): number {
    const c = UNIVERSE_CONSTANTS.physics.SPEED_OF_LIGHT_MPS;
    return (
      (c * deltaSeconds) ** 2 -
      spatialDistanceSquared(dxMeters, dyMeters, dzMeters)
    );
  },
  properTimeSeconds(
    deltaSeconds: number,
    dxMeters: number,
    dyMeters: number = 0,
    dzMeters: number = 0
  ): number {
    const interval = this.spacetimeIntervalSquared(
      deltaSeconds,
      dxMeters,
      dyMeters,
      dzMeters
    );
    return interval <= 0
      ? 0
      : Math.sqrt(interval) / UNIVERSE_CONSTANTS.physics.SPEED_OF_LIGHT_MPS;
  },
  lightTravelTimeSeconds(distanceMeters: number): number {
    return distanceMeters / UNIVERSE_CONSTANTS.physics.SPEED_OF_LIGHT_MPS;
  },
  distanceFromLightTimeMeters(seconds: number): number {
    return seconds * UNIVERSE_CONSTANTS.physics.SPEED_OF_LIGHT_MPS;
  },
  dimensionStride(...coordinates: number[]): number {
    return Math.sqrt(
      coordinates.reduce((sum, value) => sum + value * value, 0)
    );
  },
  dimensionDeficit(dimensions: number, channels: number = 1): number {
    return Math.max(0, dimensions - channels);
  },
  semioticDeficit(dimensions: number, streams: number = 1): number {
    return Math.max(0, dimensions - streams);
  },
  informationAscentBitsFromMass(
    massKg: number,
    temperatureKelvin: number = 1
  ): number {
    const joules = this.energyMassEquivalence(massKg);
    const perBit = this.landauerHeat(1, temperatureKelvin);
    return perBit === 0 ? Number.POSITIVE_INFINITY : joules / perBit;
  },
  qubitProbability(qubit: QuantumLiteral, basis: '0' | '1'): number {
    return basis === '0' ? qubit.alpha ** 2 : qubit.beta ** 2;
  },

  // ── storms.watch consolidation: void weather through the god formula ──

  /** Void net mass from god formula weight.
   *  voidNetMass = godWeight(volume, repulsion) - repulsion - 1
   *  Proved in Lean: consolidation_void_net_from_god */
  voidNetMass(volume: number, repulsion: number): number {
    return this.godFormulaWeight(volume, repulsion) - repulsion - 1;
  },

  /** Void skill = void net mass. The sufficient statistic for forecast quality.
   *  Proved in Lean: storms_skill_eq_void_net_mass */
  voidSkill(batna: number, watna: number, nullCore: number): number {
    return batna - watna - nullCore;
  },

  /** Vulnerability penalty = batna * watna. The price of internal contradiction.
   *  Zero for pure voids. Quadratic for mixed.
   *  Proved in Lean: storms_vulnerability_decomposition */
  voidPenalty(batna: number, watna: number): number {
    return batna * watna;
  },

  /** Proper skill = skill - penalty. Honest forecast quality.
   *  Proved in Lean: storms_proper_skill_eq (concrete witnesses) */
  properSkill(batna: number, watna: number, nullCore: number): number {
    return (
      this.voidSkill(batna, watna, nullCore) - this.voidPenalty(batna, watna)
    );
  },

  /** The one-competitor theorem: with w=1, proper skill = -1 for all b.
   *  Proved in Lean: storms_one_competitor_theorem */
  oneCompetitorCost(): number {
    return -1;
  },

  /** Monopoly skill: with w=0, proper skill = batna.
   *  Proved in Lean: storms_monopoly */
  monopolySkill(batna: number): number {
    return batna;
  },
};

const SHA256_INITIAL_STATE = [
  0x6a09e667, 0xbb67ae85, 0x3c6ef372, 0xa54ff53a, 0x510e527f, 0x9b05688c,
  0x1f83d9ab, 0x5be0cd19,
];

const SHA256_CONSTANTS = [
  0x428a2f98, 0x71374491, 0xb5c0fbcf, 0xe9b5dba5, 0x3956c25b, 0x59f111f1,
  0x923f82a4, 0xab1c5ed5, 0xd807aa98, 0x12835b01, 0x243185be, 0x550c7dc3,
  0x72be5d74, 0x80deb1fe, 0x9bdc06a7, 0xc19bf174, 0xe49b69c1, 0xefbe4786,
  0x0fc19dc6, 0x240ca1cc, 0x2de92c6f, 0x4a7484aa, 0x5cb0a9dc, 0x76f988da,
  0x983e5152, 0xa831c66d, 0xb00327c8, 0xbf597fc7, 0xc6e00bf3, 0xd5a79147,
  0x06ca6351, 0x14292967, 0x27b70a85, 0x2e1b2138, 0x4d2c6dfc, 0x53380d13,
  0x650a7354, 0x766a0abb, 0x81c2c92e, 0x92722c85, 0xa2bfe8a1, 0xa81a664b,
  0xc24b8b70, 0xc76c51a3, 0xd192e819, 0xd6990624, 0xf40e3585, 0x106aa070,
  0x19a4c116, 0x1e376c08, 0x2748774c, 0x34b0bcb5, 0x391c0cb3, 0x4ed8aa4a,
  0x5b9cca4f, 0x682e6ff3, 0x748f82ee, 0x78a5636f, 0x84c87814, 0x8cc70208,
  0x90befffa, 0xa4506ceb, 0xbef9a3f7, 0xc67178f2,
];

const textEncoder = new TextEncoder();

function rotateRight(value: number, shift: number): number {
  return (value >>> shift) | (value << (32 - shift));
}

function add32(...values: number[]): number {
  let sum = 0;
  for (const value of values) {
    sum = (sum + value) >>> 0;
  }
  return sum;
}

function sha256(bytes: Uint8Array): Uint8Array {
  const bitLength = bytes.length * 8;
  const paddingBytes = (64 - ((bytes.length + 9) % 64)) % 64;
  const padded = new Uint8Array(bytes.length + 1 + paddingBytes + 8);
  padded.set(bytes);
  padded[bytes.length] = 0x80;

  const bitLengthHigh = Math.floor(bitLength / 0x100000000);
  const bitLengthLow = bitLength >>> 0;
  const view = new DataView(padded.buffer);
  view.setUint32(padded.length - 8, bitLengthHigh, false);
  view.setUint32(padded.length - 4, bitLengthLow, false);

  const hashState = [...SHA256_INITIAL_STATE];
  const schedule = new Uint32Array(64);

  for (let offset = 0; offset < padded.length; offset += 64) {
    for (let index = 0; index < 16; index += 1) {
      schedule[index] = view.getUint32(offset + index * 4, false);
    }

    for (let index = 16; index < 64; index += 1) {
      const sigma0 =
        rotateRight(schedule[index - 15]!, 7) ^
        rotateRight(schedule[index - 15]!, 18) ^
        (schedule[index - 15]! >>> 3);
      const sigma1 =
        rotateRight(schedule[index - 2]!, 17) ^
        rotateRight(schedule[index - 2]!, 19) ^
        (schedule[index - 2]! >>> 10);

      schedule[index] = add32(
        schedule[index - 16]!,
        sigma0,
        schedule[index - 7]!,
        sigma1
      );
    }

    let [a, b, c, d, e, f, g, h] = hashState;

    for (let index = 0; index < 64; index += 1) {
      const bigSigma1 =
        rotateRight(e!, 6) ^ rotateRight(e!, 11) ^ rotateRight(e!, 25);
      const choice = (e! & f!) ^ (~e! & g!);
      const temp1 = add32(
        h!,
        bigSigma1,
        choice,
        SHA256_CONSTANTS[index]!,
        schedule[index]!
      );
      const bigSigma0 =
        rotateRight(a!, 2) ^ rotateRight(a!, 13) ^ rotateRight(a!, 22);
      const majority = (a! & b!) ^ (a! & c!) ^ (b! & c!);
      const temp2 = add32(bigSigma0, majority);

      h = g;
      g = f;
      f = e;
      e = add32(d!, temp1);
      d = c;
      c = b;
      b = a;
      a = add32(temp1, temp2);
    }

    hashState[0] = add32(hashState[0]!, a!);
    hashState[1] = add32(hashState[1]!, b!);
    hashState[2] = add32(hashState[2]!, c!);
    hashState[3] = add32(hashState[3]!, d!);
    hashState[4] = add32(hashState[4]!, e!);
    hashState[5] = add32(hashState[5]!, f!);
    hashState[6] = add32(hashState[6]!, g!);
    hashState[7] = add32(hashState[7]!, h!);
  }

  const digest = new Uint8Array(32);
  const digestView = new DataView(digest.buffer);
  for (let index = 0; index < hashState.length; index += 1) {
    digestView.setUint32(index * 4, hashState[index]!, false);
  }
  return digest;
}

/**
 * Handles the gnosis stable Stringify workflow.
 */
export function stableStringify(value: unknown): string {
  if (value === undefined) {
    return 'null';
  }

  if (Array.isArray(value)) {
    return `[${value.map((entry) => stableStringify(entry)).join(',')}]`;
  }

  if (value && typeof value === 'object') {
    const entries = Object.entries(value as Record<string, unknown>).sort(
      ([left], [right]) => left.localeCompare(right)
    );
    return `{${entries
      .map(([key, entry]) => `${JSON.stringify(key)}:${stableStringify(entry)}`)
      .join(',')}}`;
  }

  return JSON.stringify(value);
}

/**
 * Handles the gnosis sha256 Hex workflow.
 */
export function sha256Hex(value: string): string {
  const digest = sha256(textEncoder.encode(value));
  return Array.from(digest, (byte) => byte.toString(16).padStart(2, '0')).join(
    ''
  );
}

/**
 * Computes the Law Program Metrics.
 */
export function computeLawProgramMetrics(
  program: LawProgram
): LawProgramMetrics {
  let dimensionCount = 0;
  for (const edge of program.edges) {
    if (
      edge.type === 'FORK' ||
      edge.type === 'EVOLVE' ||
      edge.type === 'SUPERPOSE' ||
      edge.type === 'ENTANGLE'
    ) {
      dimensionCount++;
    }
  }

  return {
    nodeCount: program.nodes.size,
    edgeCount: program.edges.length,
    dimensionCount: Math.max(1, dimensionCount),
  };
}

/**
 * Handles the gnosis canonicalize Law Program workflow.
 */
export function canonicalizeLawProgram(program: LawProgram): string {
  const nodes = [...program.nodes.values()]
    .map((node) => ({
      id: node.id,
      labels: [...node.labels].sort(),
      properties: Object.fromEntries(
        Object.entries(node.properties).sort(([left], [right]) =>
          left.localeCompare(right)
        )
      ),
    }))
    .sort((left, right) => left.id.localeCompare(right.id));

  const edges = [...program.edges]
    .map((edge) => ({
      sourceIds: [...edge.sourceIds],
      targetIds: [...edge.targetIds],
      type: edge.type,
      properties: Object.fromEntries(
        Object.entries(edge.properties).sort(([left], [right]) =>
          left.localeCompare(right)
        )
      ),
    }))
    .sort((left, right) =>
      stableStringify(left).localeCompare(stableStringify(right))
    );

  return stableStringify({
    nodes,
    edges,
    subject: program.subject ?? 'topology',
    metadata: program.metadata ?? {},
  });
}

/**
 * Handles the gnosis digest Law Program workflow.
 */
export function digestLawProgram(program: LawProgram): string {
  return sha256Hex(canonicalizeLawProgram(program));
}

/**
 * Creates the Law State.
 */
export function createLawState(certificate: UniverseCertificate): LawState {
  return {
    kernelId: certificate.kernelId,
    status: certificate.status,
    dimension: certificate.dimension,
    currentDimension: certificate.currentDimension,
    deficit: certificate.clockwork.deficit.total,
    heat: certificate.clockwork.heat.totalHeat,
    conservation: certificate.clockwork.conservation.conserved,
    certificateDigest: certificate.digest,
    theoremRefs: certificate.theoremRefs,
    certificate,
    constants: UNIVERSE_CONSTANTS,
    formulas: UNIVERSE_FORMULAS,
    quantum: QUANTUM_LITERALS,
  };
}
