// @a0n/.ignored_gnosis/src/universe/universal-epoch-clock.ts
import {
  CLOCK_MAX_INDEX,
  PELL_TOWER,
  TRIB_TOWER,
  vickreyMoment,
} from '../quasicrystalline-clock';
import { pisotF } from '../betti/pisot-primitive';

/**
 * Canonical cosmic baseline for Gnosis universe time.
 * We keep this fixed so all clocks derive from one deterministic epoch model.
 */
export const UNIVERSAL_EPOCH_REFERENCE_AGE_YEARS = 13_800_000_000n;

/**
 * Mean tropical (solar) year in exact SI seconds.
 * Using a fixed scalar keeps conversion stable across runtimes.
 */
export const SOLAR_YEAR_SECONDS = 31_556_952n;

/**
 * Universe age projection at Unix epoch (1970-01-01T00:00:00Z).
 * This is the deterministic offset used for "seconds since Universal Epoch".
 */
export const UNIVERSAL_EPOCH_SECONDS_AT_UNIX_EPOCH =
  UNIVERSAL_EPOCH_REFERENCE_AGE_YEARS * SOLAR_YEAR_SECONDS;

export interface GnosisQuasicrystallineVectorClock {
  readonly phi: string;
  readonly pell: string;
  readonly trib: string;
  readonly moment: string;
  readonly phiIndex: number;
  readonly pellIndex: number;
  readonly tribIndex: number;
}

export interface GnosisUniversalEpochClock {
  readonly epoch: 'universal-epoch';
  readonly unixSeconds: string;
  readonly secondsSinceUniversalEpoch: string;
  readonly vectorClock: GnosisQuasicrystallineVectorClock;
}

function normalizeUnixSeconds(unixMilliseconds: number): bigint {
  if (!Number.isFinite(unixMilliseconds)) {
    throw new Error('unixMilliseconds must be a finite number');
  }
  return BigInt(Math.trunc(unixMilliseconds / 1000));
}

function positiveModulo(value: bigint, divisor: bigint): bigint {
  const remainder = value % divisor;
  return remainder >= 0n ? remainder : remainder + divisor;
}

function indexFromSeed(seed: bigint, laneOffset: bigint): number {
  const span = BigInt(CLOCK_MAX_INDEX + 1);
  return Number(positiveModulo(seed + laneOffset, span));
}

/**
 * Convert a scalar second-count into the 3-lane quasicrystalline projection.
 * Lane offsets keep the three indices decorrelated while still deterministic.
 */
export function quasicrystallineVectorClock(
  secondsSinceUniversalEpochValue: bigint
): GnosisQuasicrystallineVectorClock {
  const phiIndex = indexFromSeed(secondsSinceUniversalEpochValue, 0n);
  const pellIndex = indexFromSeed(secondsSinceUniversalEpochValue / 2n, 11n);
  const tribIndex = indexFromSeed(secondsSinceUniversalEpochValue / 3n, 23n);
  const pellValue = PELL_TOWER[pellIndex];
  const tribValue = TRIB_TOWER[tribIndex];
  if (pellValue === undefined || tribValue === undefined) {
    throw new RangeError('quasicrystalline vector index out of bounds');
  }
  const moment = vickreyMoment({
    phi: phiIndex,
    pell: pellIndex,
    trib: tribIndex,
  });

  return {
    phi: String(pisotF(phiIndex)),
    pell: pellValue.toString(),
    trib: tribValue.toString(),
    moment: moment.toString(),
    phiIndex,
    pellIndex,
    tribIndex,
  };
}

/**
 * Deterministic universe-time primitive:
 * seconds elapsed since Universal Epoch = fixed cosmic offset + Unix seconds.
 */
export function secondsSinceUniversalEpoch(
  unixMilliseconds: number = Date.now()
): bigint {
  const unixSeconds = normalizeUnixSeconds(unixMilliseconds);
  return UNIVERSAL_EPOCH_SECONDS_AT_UNIX_EPOCH + unixSeconds;
}

/**
 * Structured Gnosis clock payload suitable for transport and logging.
 */
export function gnosisUniversalEpochClock(
  unixMilliseconds: number = Date.now()
): GnosisUniversalEpochClock {
  const unixSeconds = normalizeUnixSeconds(unixMilliseconds);
  const cosmicSeconds = UNIVERSAL_EPOCH_SECONDS_AT_UNIX_EPOCH + unixSeconds;
  return {
    epoch: 'universal-epoch',
    unixSeconds: unixSeconds.toString(),
    secondsSinceUniversalEpoch: cosmicSeconds.toString(),
    vectorClock: quasicrystallineVectorClock(cosmicSeconds),
  };
}
