// @a0n/.ignored_gnosis/src/universe/verdict.ts
/**
 * verdict.ts -- Clockwork verification verdicts
 *
 * A verdict is the output of self-verification: the void boundary
 * has been read, the deficit checked, conservation confirmed,
 * and the seven laws audited. The model knows its own truth state.
 *
 * Verdicts are deterministic. Same void boundary, same verdict.
 * No oracle consulted. The void did the work.
 */

// ============================================================================
// Verdict types
// ============================================================================

/** The three possible verification states */
export type VerdictStatus =
  | 'verified' // β₁ = 0 at all terminals, all laws hold
  | 'open' // β₁ > 0 somewhere, work remains
  | 'violated'; // structural invariant broken

/** A single check within the verdict */
export interface VerdictCheck {
  readonly id: string;
  readonly name: string;
  readonly pass: boolean;
  readonly message: string;
  /** Which theorem family backs this check */
  readonly theorem: string;
}

/** Conservation accounting: V_fork = W_fold + Q_vent */
export interface ConservationLedger {
  /** Total value entering via forks */
  readonly vFork: number;
  /** Useful work preserved through folds */
  readonly wFold: number;
  /** Heat dissipated through vents */
  readonly qVent: number;
  /** Conservation residual (should be 0) */
  readonly residual: number;
  /** Does the first law hold? */
  readonly conserved: boolean;
}

/** Deficit report: per-node β₁ */
export interface DeficitSummary {
  /** Total topological deficit across all nodes */
  readonly totalDeficit: number;
  /** Maximum deficit at any single node */
  readonly maxDeficit: number;
  /** Number of nodes with deficit > 0 */
  readonly openNodes: number;
  /** Total nodes checked */
  readonly totalNodes: number;
  /** Total vented (Law 5 conservation) */
  readonly totalVented: number;
  /** Per-node deficit values */
  readonly perNode: ReadonlyMap<string, number>;
}

/** Void boundary snapshot at verification time */
export interface VoidSnapshot {
  /** Per-dimension rejection counts */
  readonly counts: readonly number[];
  /** Total entries in the void */
  readonly totalEntries: number;
  /** Number of dimensions */
  readonly dimensions: number;
  /** Teleportation deficit (one integer that captures the full trajectory) */
  readonly teleportDeficit: number;
  /** Buleyean positivity: all weights >= 1 */
  readonly positivityHolds: boolean;
}

/** Landauer heat accounting */
export interface HeatLedger {
  /** Total irreversible heat in normalized kT ln 2 units */
  readonly totalHeat: number;
  /** Number of irreversible merge points */
  readonly mergePoints: number;
}

/** Bounded finite-state witness for Clockwork's Laplace surface */
export interface LaplaceWitness {
  /** Number of observed rounds/rejections backing the witness */
  readonly rounds: number;
  /** The sliver floor: every weight stays at or above 1 */
  readonly floor: 1;
  /** The finite observation ceiling: rounds + 1 */
  readonly ceiling: number;
  /** Maximum discrimination/gain available from the observed rounds */
  readonly gainCapacity: number;
  /** Concrete weight vector induced by the current void boundary */
  readonly weights: readonly number[];
  /** The void boundary remains a finite natural-number vector */
  readonly finiteBoundary: boolean;
  /** Every weight stays in the computable natural-number zone */
  readonly computableWeights: boolean;
  /** The verification surface is a finite conjunction, not a universal one */
  readonly finiteConjunction: boolean;
  /** The witness remains inside the finite-stage theorem surface */
  readonly finiteStages: boolean;
  /** Every weight stays inside the floor/ceiling sandwich */
  readonly withinZone: boolean;
}

/** The full Clockwork verdict */
export interface ClockworkVerdict {
  /** Overall status */
  readonly status: VerdictStatus;
  /** Human-readable summary */
  readonly summary: string;
  /** Void boundary state at verification time */
  readonly void: VoidSnapshot;
  /** Topological deficit analysis */
  readonly deficit: DeficitSummary;
  /** First law conservation check */
  readonly conservation: ConservationLedger;
  /** Landauer heat accounting */
  readonly heat: HeatLedger;
  /** Bounded finite-state witness for the Laplace surface */
  readonly laplace?: LaplaceWitness;
  /** Individual verification checks */
  readonly checks: readonly VerdictCheck[];
  /** Number of checks that passed */
  readonly passed: number;
  /** Number of checks that failed */
  readonly failed: number;
  /** Timestamp of verification */
  readonly verifiedAt: number;
}

// ============================================================================
// Verdict construction helpers
// ============================================================================

/**
 * Handles the gnosis pass Check workflow.
 */
export function passCheck(
  id: string,
  name: string,
  theorem: string
): VerdictCheck {
  return { id, name, pass: true, message: 'OK', theorem };
}

/**
 * Handles the gnosis fail Check workflow.
 */
export function failCheck(
  id: string,
  name: string,
  theorem: string,
  message: string
): VerdictCheck {
  return { id, name, pass: false, message, theorem };
}

/**
 * Handles the gnosis verdict Status workflow.
 */
export function verdictStatus(checks: readonly VerdictCheck[]): VerdictStatus {
  const hasViolation = checks.some((c) => !c.pass);
  if (!hasViolation) return 'verified';
  // Distinguish open (deficit > 0 but structurally sound) from violated
  const structuralFailures = checks.filter(
    (c) => !c.pass && !c.id.startsWith('deficit-')
  );
  return structuralFailures.length > 0 ? 'violated' : 'open';
}

/**
 * Handles the gnosis verdict Summary workflow.
 */
export function verdictSummary(
  status: VerdictStatus,
  deficit: DeficitSummary,
  conservation: ConservationLedger
): string {
  switch (status) {
    case 'verified':
      return `Verified: β₁ = 0 at all ${deficit.totalNodes} nodes, conservation holds, void boundary is sufficient statistic`;
    case 'open':
      return `Open: β₁ > 0 at ${deficit.openNodes}/${deficit.totalNodes} nodes (total deficit = ${deficit.totalDeficit}), ${deficit.totalDeficit} Bules of work remain`;
    case 'violated':
      return `Violated: structural invariant broken (conservation residual = ${conservation.residual})`;
  }
}
