// @a0n/gnosis/src/socio-topology/rollback-controller.ts
/**
 * Quarantine rollback policy: integral deficit on λ₂ hat + consecutive low + β₁ floor.
 * Pure functions; no I/O. Aligns with docs/TOPO_COGNITIVE_EVENT_HORIZON.md.
 */

export interface RollbackPolicy {
  readonly lambda2Star: number;
  /** Consecutive windows below λ₂* */
  readonly tauWindows: number;
  readonly integralBudgetW: number;
  readonly beta1Min: number;
}

export interface RollbackSample {
  readonly lambda2Hat: number;
  readonly beta1Hat: number;
  readonly t: number;
}

export type RollbackDecision =
  | { action: 'hold' }
  | { action: 'rollback'; reason: 'consecutive_low' | 'integral_deficit' | 'betti_floor' };

export class RollbackController {
  private readonly policy: RollbackPolicy;
  private deficitIntegral = 0;
  private consecutiveLow = 0;
  private lastT = -1;

  constructor(policy: RollbackPolicy) {
    this.policy = policy;
  }

  reset(): void {
    this.deficitIntegral = 0;
    this.consecutiveLow = 0;
    this.lastT = -1;
  }

  /** Feed one metric sample; returns rollback decision. */
  step(s: RollbackSample): RollbackDecision {
    const { lambda2Star, tauWindows, integralBudgetW, beta1Min } = this.policy;
    if (s.beta1Hat < beta1Min) {
      return { action: 'rollback', reason: 'betti_floor' };
    }
    if (s.lambda2Hat < lambda2Star) {
      this.consecutiveLow += 1;
      this.deficitIntegral += lambda2Star - s.lambda2Hat;
    } else {
      this.consecutiveLow = 0;
    }
    if (this.consecutiveLow >= tauWindows) {
      return { action: 'rollback', reason: 'consecutive_low' };
    }
    if (this.deficitIntegral > integralBudgetW) {
      return { action: 'rollback', reason: 'integral_deficit' };
    }
    this.lastT = s.t;
    return { action: 'hold' };
  }

  getDeficitIntegral(): number {
    return this.deficitIntegral;
  }
}
