// @a0n/.ignored_gnosis/harnesses/node_framework_compat.mjs
import fs from 'node:fs';
import path from 'node:path';
import { pathToFileURL } from 'node:url';

const REGISTRY_KEY = '__xgnosis_node_router_registry__';

function registry() {
  if (!Array.isArray(globalThis[REGISTRY_KEY])) {
    globalThis[REGISTRY_KEY] = [];
  }

  return globalThis[REGISTRY_KEY];
}

function clearRegistry() {
  globalThis[REGISTRY_KEY] = [];
}

function normalizePathname(value) {
  if (!value || value === '*') {
    return '/';
  }

  const [pathname] = String(value).split('?', 1);
  if (!pathname || pathname === '') {
    return '/';
  }

  return pathname.startsWith('/') ? pathname : `/${pathname}`;
}

function normalizeQuery(value) {
  if (!value) {
    return '';
  }

  return value.startsWith('?') ? value : `?${value}`;
}

function trimTrailingSlash(value) {
  if (value.length > 1 && value.endsWith('/')) {
    return value.slice(0, -1);
  }

  return value;
}

function joinMountedPath(prefix, childPath) {
  if (!prefix || prefix === '*' || prefix === '/*') {
    return childPath;
  }

  const normalizedPrefix = trimTrailingSlash(normalizePathname(prefix));
  const normalizedChild =
    childPath === '*' || childPath === '/*'
      ? ''
      : trimTrailingSlash(normalizePathname(childPath));

  if (!normalizedChild || normalizedChild === '/') {
    return normalizedPrefix;
  }

  return `${normalizedPrefix}${normalizedChild}`;
}

function escapeRegExp(value) {
  return value.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

function compileRoutePattern(pattern) {
  if (pattern === '*' || pattern === '/*') {
    return {
      regex: /^\/.*$/,
      keys: [],
    };
  }

  const normalized = normalizePathname(pattern);
  if (normalized === '/') {
    return {
      regex: /^\/$/,
      keys: [],
    };
  }

  const keys = [];
  const segments = normalized
    .split('/')
    .filter((segment) => segment.length > 0)
    .map((segment) => {
      if (segment.startsWith(':')) {
        const key = segment.slice(1);
        keys.push(key);
        return '([^/]+)';
      }

      if (segment.startsWith('*')) {
        const key = segment.slice(1) || 'wildcard';
        keys.push(key);
        return '(.*)';
      }

      if (segment === '*') {
        keys.push('wildcard');
        return '(.*)';
      }

      return escapeRegExp(segment);
    });

  return {
    regex: new RegExp(`^/${segments.join('/')}/?$`),
    keys,
  };
}

function matchRoute(pattern, pathname) {
  const compiled = compileRoutePattern(pattern);
  const match = compiled.regex.exec(normalizePathname(pathname));
  if (!match) {
    return null;
  }

  const params = {};
  for (let index = 0; index < compiled.keys.length; index += 1) {
    params[compiled.keys[index]] = decodeURIComponent(match[index + 1] ?? '');
  }

  return params;
}

function matchMiddlewarePrefix(prefix, pathname) {
  if (!prefix || prefix === '*' || prefix === '/*') {
    return true;
  }

  const normalizedPath = normalizePathname(pathname);
  const normalizedPrefix = normalizePathname(prefix).replace(/\/\*$/, '');

  if (normalizedPrefix === '/') {
    return true;
  }

  return (
    normalizedPath === normalizedPrefix ||
    normalizedPath.startsWith(`${normalizedPrefix}/`)
  );
}

function queryRecord(search) {
  const params = new URLSearchParams(
    search.startsWith('?') ? search.slice(1) : search
  );
  const result = {};
  for (const [key, value] of params.entries()) {
    if (Object.prototype.hasOwnProperty.call(result, key)) {
      const current = result[key];
      result[key] = Array.isArray(current)
        ? [...current, value]
        : [current, value];
      continue;
    }
    result[key] = value;
  }
  return result;
}

function decodeBody(body) {
  if (!body || body.encoding !== 'base64' || typeof body.value !== 'string') {
    return Buffer.alloc(0);
  }

  return Buffer.from(body.value, 'base64');
}

function serializeBody(buffer) {
  return {
    encoding: 'base64',
    value: Buffer.from(buffer ?? Buffer.alloc(0)).toString('base64'),
  };
}

function normalizeResponseHeaders(headers) {
  const normalized = {};
  for (const [key, value] of Object.entries(headers ?? {})) {
    if (typeof value === 'string') {
      normalized[key] = value;
      continue;
    }

    if (Array.isArray(value)) {
      normalized[key] = value.join(', ');
      continue;
    }

    if (value !== undefined && value !== null) {
      normalized[key] = String(value);
    }
  }

  return normalized;
}

function createResponseState() {
  return {
    status: 200,
    headers: {},
    body: Buffer.alloc(0),
    finished: false,
  };
}

function setBody(state, body, contentType) {
  if (body === undefined || body === null) {
    state.body = Buffer.alloc(0);
  } else if (Buffer.isBuffer(body)) {
    state.body = body;
  } else if (body instanceof Uint8Array) {
    state.body = Buffer.from(body);
  } else if (typeof body === 'string') {
    state.body = Buffer.from(body, 'utf8');
  } else {
    state.body = Buffer.from(JSON.stringify(body), 'utf8');
    if (!state.headers['Content-Type'] && !state.headers['content-type']) {
      state.headers['Content-Type'] = 'application/json; charset=utf-8';
    }
  }

  if (contentType) {
    state.headers['Content-Type'] = contentType;
  }

  state.finished = true;
}

function finalizeResponse(state) {
  return {
    matched: true,
    response: {
      status: state.status,
      headers: normalizeResponseHeaders(state.headers),
      body: serializeBody(state.body),
    },
  };
}

function isExpressBuiltinJson(handler) {
  return (
    handler &&
    typeof handler === 'object' &&
    handler.__xgnosisBuiltin === 'express.json'
  );
}

function isExpressBuiltinStatic(handler) {
  return (
    handler &&
    typeof handler === 'object' &&
    handler.__xgnosisBuiltin === 'express.static'
  );
}

function maybeParseJsonBody(request, responseState) {
  const contentType = request.headers['content-type'] ?? '';
  if (!contentType.toLowerCase().includes('application/json')) {
    return;
  }

  const text = request.rawBody.toString('utf8').trim();
  if (text.length === 0) {
    request.body = {};
    return;
  }

  try {
    request.body = JSON.parse(text);
  } catch (error) {
    responseState.status = 400;
    responseState.headers['Content-Type'] = 'application/json; charset=utf-8';
    responseState.body = Buffer.from(
      JSON.stringify({
        error:
          error instanceof Error ? error.message : 'Failed to parse JSON body.',
      }),
      'utf8'
    );
    responseState.finished = true;
  }
}

function createExpressRequest(invocation, params) {
  const query = queryRecord(invocation.query ?? '');
  const rawBody = decodeBody(invocation.body);
  const originalUrl = `${invocation.path}${normalizeQuery(
    invocation.query ?? ''
  )}`;
  const headers = Object.fromEntries(
    Object.entries(invocation.headers ?? {}).map(([key, value]) => [
      key.toLowerCase(),
      value,
    ])
  );

  return {
    method: invocation.method,
    path: invocation.path,
    originalUrl,
    url: originalUrl,
    params,
    query,
    headers,
    rawBody,
    body: undefined,
    get(name) {
      return headers[String(name).toLowerCase()];
    },
    header(name) {
      return headers[String(name).toLowerCase()];
    },
  };
}

function createExpressResponse(state) {
  return {
    locals: {},
    get headersSent() {
      return state.finished;
    },
    get finished() {
      return state.finished;
    },
    get statusCode() {
      return state.status;
    },
    set statusCode(value) {
      state.status = Number(value) || 200;
    },
    status(code) {
      state.status = Number(code) || 200;
      return this;
    },
    set(name, value) {
      state.headers[String(name)] = String(value);
      return this;
    },
    header(name, value) {
      if (value === undefined) {
        return state.headers[String(name)];
      }
      state.headers[String(name)] = String(value);
      return this;
    },
    setHeader(name, value) {
      state.headers[String(name)] = String(value);
    },
    getHeader(name) {
      return state.headers[String(name)];
    },
    type(value) {
      state.headers['Content-Type'] = String(value);
      return this;
    },
    json(payload) {
      if (!state.headers['Content-Type']) {
        state.headers['Content-Type'] = 'application/json; charset=utf-8';
      }
      setBody(state, JSON.stringify(payload), state.headers['Content-Type']);
      return this;
    },
    send(payload) {
      setBody(state, payload);
      return this;
    },
    end(payload) {
      setBody(state, payload);
      return this;
    },
    redirect(location, status = 302) {
      state.status = Number(status) || 302;
      state.headers.Location = String(location);
      state.body = Buffer.alloc(0);
      state.finished = true;
      return this;
    },
    sendStatus(code) {
      state.status = Number(code) || 200;
      state.body = Buffer.alloc(0);
      state.finished = true;
      return this;
    },
  };
}

async function executeExpressChain(
  handlers,
  request,
  response,
  state,
  index = 0
) {
  if (state.finished || index >= handlers.length) {
    return;
  }

  const handler = handlers[index];

  if (isExpressBuiltinJson(handler)) {
    maybeParseJsonBody(request, state);
    if (!state.finished) {
      await executeExpressChain(handlers, request, response, state, index + 1);
    }
    return;
  }

  if (isExpressBuiltinStatic(handler)) {
    await executeExpressChain(handlers, request, response, state, index + 1);
    return;
  }

  if (typeof handler !== 'function') {
    await executeExpressChain(handlers, request, response, state, index + 1);
    return;
  }

  let nextCalled = false;
  const next = async (error) => {
    if (error) {
      throw error;
    }
    nextCalled = true;
    await executeExpressChain(handlers, request, response, state, index + 1);
  };

  if (handler.length >= 4) {
    await next();
    return;
  }

  await handler(request, response, next);

  if (!state.finished && nextCalled) {
    return;
  }

  if (!state.finished && index < handlers.length - 1) {
    await executeExpressChain(handlers, request, response, state, index + 1);
  }
}

async function dispatchExpressApp(app, invocation) {
  const pathname = normalizePathname(invocation.path);
  const route = app.__xgnosis.routes.find((candidate) => {
    if (
      candidate.method !== 'ALL' &&
      candidate.method !== String(invocation.method).toUpperCase()
    ) {
      return false;
    }

    return matchRoute(candidate.path, pathname) !== null;
  });

  if (!route) {
    return { matched: false };
  }

  const params = matchRoute(route.path, pathname) ?? {};
  const request = createExpressRequest(invocation, params);
  const state = createResponseState();
  const response = createExpressResponse(state);
  const middleware = app.__xgnosis.middleware
    .filter((candidate) => matchMiddlewarePrefix(candidate.path, pathname))
    .flatMap((candidate) => candidate.handlers);
  const handlers = [...middleware, ...route.handlers];

  await executeExpressChain(handlers, request, response, state);

  if (!state.finished) {
    throw new Error(
      `Express compatibility handler for ${invocation.method} ${pathname} completed without sending a response.`
    );
  }

  return finalizeResponse(state);
}

function createHonoResponseRecord(state) {
  return {
    __xgnosisHonoResponse: true,
    status: state.status,
    headers: { ...state.headers },
    body: Buffer.from(state.body),
  };
}

function applyHonoResponseRecord(state, record) {
  state.status = record.status;
  state.headers = { ...record.headers };
  state.body = Buffer.from(record.body);
  state.finished = true;
}

function createHonoContext(invocation, params, state) {
  const query = queryRecord(invocation.query ?? '');
  const rawBody = decodeBody(invocation.body);
  const headers = Object.fromEntries(
    Object.entries(invocation.headers ?? {}).map(([key, value]) => [
      key.toLowerCase(),
      value,
    ])
  );
  const url = `http://${headers.host ?? 'localhost'}${
    invocation.path
  }${normalizeQuery(invocation.query ?? '')}`;

  const context = {
    env: {},
    executionCtx: undefined,
    req: {
      method: invocation.method,
      path: invocation.path,
      url,
      rawBody,
      header(name) {
        if (name === undefined) {
          return { ...headers };
        }
        return headers[String(name).toLowerCase()];
      },
      param(name) {
        if (name === undefined) {
          return { ...params };
        }
        return params[String(name)];
      },
      query(name) {
        if (name === undefined) {
          return { ...query };
        }
        return query[String(name)];
      },
      async json() {
        const text = rawBody.toString('utf8').trim();
        return text.length === 0 ? {} : JSON.parse(text);
      },
      async text() {
        return rawBody.toString('utf8');
      },
    },
    status(code) {
      state.status = Number(code) || 200;
      return context;
    },
    header(name, value) {
      state.headers[String(name)] = String(value);
      return context;
    },
    json(payload, status) {
      if (status !== undefined) {
        state.status = Number(status) || 200;
      }
      state.headers['Content-Type'] = 'application/json; charset=utf-8';
      state.body = Buffer.from(JSON.stringify(payload), 'utf8');
      state.finished = true;
      return createHonoResponseRecord(state);
    },
    text(payload, status) {
      if (status !== undefined) {
        state.status = Number(status) || 200;
      }
      state.headers['Content-Type'] = 'text/plain; charset=utf-8';
      state.body = Buffer.from(String(payload), 'utf8');
      state.finished = true;
      return createHonoResponseRecord(state);
    },
    body(payload, status) {
      if (status !== undefined) {
        state.status = Number(status) || 200;
      }
      setBody(state, payload);
      return createHonoResponseRecord(state);
    },
  };

  return context;
}

async function executeHonoChain(handlers, context, state, index = 0) {
  if (state.finished || index >= handlers.length) {
    return undefined;
  }

  const handler = handlers[index];
  if (typeof handler !== 'function') {
    return executeHonoChain(handlers, context, state, index + 1);
  }

  let nextCalled = false;
  const next = async () => {
    nextCalled = true;
    return executeHonoChain(handlers, context, state, index + 1);
  };

  const result = await handler(context, next);
  if (result && result.__xgnosisHonoResponse) {
    applyHonoResponseRecord(state, result);
    return result;
  }

  if (!nextCalled && !state.finished && index < handlers.length - 1) {
    return executeHonoChain(handlers, context, state, index + 1);
  }

  return result;
}

async function dispatchHonoApp(app, invocation) {
  const pathname = normalizePathname(invocation.path);
  const route = app.__xgnosis.routes.find((candidate) => {
    if (
      candidate.method !== 'ALL' &&
      candidate.method !== String(invocation.method).toUpperCase()
    ) {
      return false;
    }

    return matchRoute(candidate.path, pathname) !== null;
  });

  if (!route) {
    return { matched: false };
  }

  const params = matchRoute(route.path, pathname) ?? {};
  const state = createResponseState();
  const context = createHonoContext(invocation, params, state);
  const middleware = app.__xgnosis.middleware
    .filter((candidate) => matchMiddlewarePrefix(candidate.path, pathname))
    .flatMap((candidate) => candidate.handlers);
  const handlers = [...middleware, ...route.handlers];
  const result = await executeHonoChain(handlers, context, state);

  if (!state.finished) {
    if (result && result.__xgnosisHonoResponse) {
      applyHonoResponseRecord(state, result);
    } else if (result !== undefined && result !== null) {
      setBody(state, result);
    }
  }

  if (!state.finished) {
    throw new Error(
      `Hono compatibility handler for ${invocation.method} ${pathname} completed without sending a response.`
    );
  }

  return finalizeResponse(state);
}

function parseCompatRequestBody(headers, rawBody) {
  const contentType = String(headers['content-type'] ?? '').toLowerCase();
  const text = rawBody.toString('utf8');
  if (contentType.includes('application/json')) {
    return text.trim().length === 0 ? {} : JSON.parse(text);
  }
  return text;
}

function createKoaContext(invocation, params, state) {
  const query = queryRecord(invocation.query ?? '');
  const rawBody = decodeBody(invocation.body);
  const headers = Object.fromEntries(
    Object.entries(invocation.headers ?? {}).map(([key, value]) => [
      key.toLowerCase(),
      value,
    ])
  );
  const url = `${invocation.path}${normalizeQuery(invocation.query ?? '')}`;
  const requestBody = parseCompatRequestBody(headers, rawBody);

  const ctx = {
    state: {},
    app: undefined,
    method: invocation.method,
    path: invocation.path,
    params,
    query,
    req: {
      method: invocation.method,
      path: invocation.path,
      url,
      headers,
      body: requestBody,
      rawBody,
    },
    request: {
      method: invocation.method,
      path: invocation.path,
      url,
      headers,
      query,
      body: requestBody,
      rawBody,
      get(name) {
        return headers[String(name).toLowerCase()];
      },
    },
    response: {
      headers: state.headers,
    },
    get headers() {
      return headers;
    },
    get status() {
      return state.status;
    },
    set status(value) {
      state.status = Number(value) || 200;
    },
    get body() {
      return state.koaBody;
    },
    set body(value) {
      state.koaBody = value;
      state.finished = true;
    },
    set(name, value) {
      state.headers[String(name)] = String(value);
      return ctx;
    },
    get(name) {
      return headers[String(name).toLowerCase()];
    },
  };

  return ctx;
}

async function executeKoaChain(handlers, ctx, index = 0) {
  if (index >= handlers.length) {
    return undefined;
  }

  const handler = handlers[index];
  if (typeof handler !== 'function') {
    return executeKoaChain(handlers, ctx, index + 1);
  }

  const result = await handler(ctx, async () =>
    executeKoaChain(handlers, ctx, index + 1)
  );
  if (result !== undefined && ctx.body === undefined) {
    ctx.body = result;
  }
  return result;
}

function finalizeKoaResponse(state) {
  if (state.koaBody !== undefined) {
    setBody(state, state.koaBody);
  } else if (!state.finished) {
    state.body = Buffer.alloc(0);
    state.finished = true;
  }

  return finalizeResponse(state);
}

async function dispatchKoaApp(app, invocation) {
  const pathname = normalizePathname(invocation.path);
  const route = app.__xgnosis.routes.find((candidate) => {
    if (
      candidate.method !== 'ALL' &&
      candidate.method !== String(invocation.method).toUpperCase()
    ) {
      return false;
    }

    return matchRoute(candidate.path, pathname) !== null;
  });

  if (!route) {
    return { matched: false };
  }

  const params = matchRoute(route.path, pathname) ?? {};
  const state = createResponseState();
  const ctx = createKoaContext(invocation, params, state);
  ctx.app = app;
  const middleware = app.__xgnosis.middleware
    .filter((candidate) => matchMiddlewarePrefix(candidate.path, pathname))
    .flatMap((candidate) => candidate.handlers);
  const handlers = [...middleware, ...route.handlers];

  await executeKoaChain(handlers, ctx);
  return finalizeKoaResponse(state);
}

function createFastifyRequest(invocation, params) {
  const rawBody = decodeBody(invocation.body);
  const headers = Object.fromEntries(
    Object.entries(invocation.headers ?? {}).map(([key, value]) => [
      key.toLowerCase(),
      value,
    ])
  );

  return {
    method: invocation.method,
    url: `${invocation.path}${normalizeQuery(invocation.query ?? '')}`,
    path: invocation.path,
    params,
    query: queryRecord(invocation.query ?? ''),
    headers,
    body:
      rawBody.length > 0 ? parseCompatRequestBody(headers, rawBody) : undefined,
    raw: {
      body: rawBody,
    },
  };
}

function createFastifyReply(state) {
  return {
    get sent() {
      return state.finished;
    },
    get statusCode() {
      return state.status;
    },
    set statusCode(value) {
      state.status = Number(value) || 200;
    },
    code(value) {
      state.status = Number(value) || 200;
      return this;
    },
    status(value) {
      state.status = Number(value) || 200;
      return this;
    },
    header(name, value) {
      state.headers[String(name)] = String(value);
      return this;
    },
    type(value) {
      state.headers['Content-Type'] = String(value);
      return this;
    },
    send(payload) {
      setBody(state, payload, state.headers['Content-Type']);
      return this;
    },
  };
}

async function callFastifyLifecycle(handler, request, reply) {
  if (typeof handler !== 'function') {
    return;
  }

  if (handler.length >= 3) {
    await new Promise((resolve, reject) => {
      let settled = false;
      const done = (error) => {
        if (settled) {
          return;
        }
        settled = true;
        if (error) {
          reject(error);
          return;
        }
        resolve();
      };

      try {
        const result = handler(request, reply, done);
        if (result && typeof result.then === 'function') {
          result.then(() => done(), reject);
        }
      } catch (error) {
        reject(error);
      }
    });
    return;
  }

  await handler(request, reply);
}

async function dispatchFastifyApp(app, invocation) {
  const pathname = normalizePathname(invocation.path);
  const route = app.__xgnosis.routes.find((candidate) => {
    if (
      candidate.method !== 'ALL' &&
      candidate.method !== String(invocation.method).toUpperCase()
    ) {
      return false;
    }

    return matchRoute(candidate.path, pathname) !== null;
  });

  if (!route) {
    return { matched: false };
  }

  const params = matchRoute(route.path, pathname) ?? {};
  const state = createResponseState();
  const request = createFastifyRequest(invocation, params);
  const reply = createFastifyReply(state);

  for (const hook of route.onRequestHooks ?? []) {
    await callFastifyLifecycle(hook, request, reply);
    if (state.finished) {
      return finalizeResponse(state);
    }
  }

  for (const hook of route.preHandlers ?? []) {
    await callFastifyLifecycle(hook, request, reply);
    if (state.finished) {
      return finalizeResponse(state);
    }
  }

  const result =
    typeof route.handler === 'function'
      ? await route.handler(request, reply)
      : undefined;
  if (!state.finished && result !== undefined) {
    reply.send(result);
  }

  if (!state.finished) {
    throw new Error(
      `Fastify compatibility handler for ${invocation.method} ${pathname} completed without sending a response.`
    );
  }

  return finalizeResponse(state);
}

function detectModuleFlavor(sourceText, filePath) {
  const extension = path.extname(filePath).toLowerCase();
  if (extension === '.cjs' || extension === '.cts') {
    return 'cjs';
  }

  if (extension === '.mjs' || extension === '.mts') {
    return 'esm';
  }

  const hasRequire = /\brequire\s*\(/.test(sourceText);
  const hasImport = /\bimport\s.+from\s+['"]/.test(sourceText);
  return hasRequire && !hasImport ? 'cjs' : 'esm';
}

function compatibilityTempExtension(filePath, moduleFlavor) {
  const extension = path.extname(filePath).toLowerCase();
  if (moduleFlavor === 'cjs') {
    if (extension === '.ts' || extension === '.tsx' || extension === '.cts') {
      return '.cts';
    }
    return '.cjs';
  }

  if (extension === '.ts' || extension === '.tsx' || extension === '.mts') {
    return extension;
  }

  if (extension === '.jsx') {
    return '.jsx';
  }

  if (extension === '.mjs') {
    return '.mjs';
  }

  return '.mjs';
}

function relativeShimSpecifier(sourceDir, shimPath, moduleFlavor) {
  const relativePath = path.relative(sourceDir, shimPath);
  const normalized =
    relativePath.startsWith('./') || relativePath.startsWith('../')
      ? relativePath
      : `./${relativePath}`;
  return normalized
    .split(path.sep)
    .join(moduleFlavor === 'cjs' ? path.sep : '/');
}

function rewritePackageImport(sourceText, packageName, shimSpecifier) {
  const escaped = escapeRegExp(packageName);
  return sourceText
    .replace(
      new RegExp(`from\\s+['"]${escaped}['"]`, 'g'),
      `from '${shimSpecifier}'`
    )
    .replace(
      new RegExp(`import\\s+['"]${escaped}['"]`, 'g'),
      `import '${shimSpecifier}'`
    )
    .replace(
      new RegExp(`require\\(['"]${escaped}['"]\\)`, 'g'),
      `require('${shimSpecifier}')`
    );
}

function frameworkShims(frameworkId, moduleFlavor) {
  if (frameworkId === 'express') {
    return [
      {
        packageName: 'express',
        source: renderExpressShim(moduleFlavor),
      },
    ];
  }

  if (frameworkId === 'hono') {
    return [
      {
        packageName: 'hono',
        source: renderHonoShim(moduleFlavor),
      },
    ];
  }

  if (frameworkId === 'koa') {
    return [
      {
        packageName: 'koa',
        source: renderKoaShim(moduleFlavor),
      },
      {
        packageName: '@koa/router',
        source: renderKoaRouterShim(moduleFlavor),
      },
      {
        packageName: 'koa-router',
        source: renderKoaRouterShim(moduleFlavor),
      },
    ];
  }

  if (frameworkId === 'fastify') {
    return [
      {
        packageName: 'fastify',
        source: renderFastifyShim(moduleFlavor),
      },
    ];
  }

  return [];
}

function renderExpressShim(moduleFlavor) {
  const source = `
const REGISTRY_KEY = ${JSON.stringify(REGISTRY_KEY)};
function registry() {
  if (!Array.isArray(globalThis[REGISTRY_KEY])) {
    globalThis[REGISTRY_KEY] = [];
  }
  return globalThis[REGISTRY_KEY];
}
function normalizePathname(value) {
  if (!value || value === '*') return '/';
  const path = String(value).split('?', 1)[0] || '/';
  return path.startsWith('/') ? path : \`/\${path}\`;
}
function trimTrailingSlash(value) {
  return value.length > 1 && value.endsWith('/') ? value.slice(0, -1) : value;
}
function joinMountedPath(prefix, childPath) {
  if (!prefix || prefix === '*' || prefix === '/*') return childPath;
  const normalizedPrefix = trimTrailingSlash(normalizePathname(prefix));
  const normalizedChild =
    childPath === '*' || childPath === '/*' ? '' : trimTrailingSlash(normalizePathname(childPath));
  if (!normalizedChild || normalizedChild === '/') return normalizedPrefix;
  return \`\${normalizedPrefix}\${normalizedChild}\`;
}
function createExpressApp(kind = 'app') {
  const record = {
    framework: 'express',
    kind,
    routes: [],
    middleware: [],
    listened: false,
  };
  const app = {
    __xgnosis: record,
    use(...args) {
      const [first, ...rest] = args;
      const hasPath = typeof first === 'string';
      const mountPath = hasPath ? first : null;
      const handlers = hasPath ? rest : args;
      for (const handler of handlers) {
        if (handler && handler.__xgnosis && handler.__xgnosis.framework === 'express') {
          for (const route of handler.__xgnosis.routes) {
            record.routes.push({
              method: route.method,
              path: joinMountedPath(mountPath, route.path),
              handlers: route.handlers,
            });
          }
          for (const middleware of handler.__xgnosis.middleware) {
            record.middleware.push({
              path: joinMountedPath(mountPath, middleware.path ?? '*'),
              handlers: middleware.handlers,
            });
          }
          continue;
        }
        record.middleware.push({
          path: mountPath,
          handlers: [handler],
        });
      }
      return app;
    },
    listen() {
      record.listened = true;
      return {
        close() {},
      };
    },
  };
  for (const method of ['get', 'post', 'put', 'delete', 'patch', 'head', 'options', 'all']) {
    app[method] = (routePath, ...handlers) => {
      record.routes.push({
        method: method.toUpperCase(),
        path: normalizePathname(routePath),
        handlers,
      });
      return app;
    };
  }
  registry().push(app);
  return app;
}
function express() {
  return createExpressApp('app');
}
express.Router = function Router() {
  return createExpressApp('router');
};
express.json = function json(options = {}) {
  return {
    __xgnosisBuiltin: 'express.json',
    options,
  };
};
express.urlencoded = function urlencoded(options = {}) {
  return {
    __xgnosisBuiltin: 'express.urlencoded',
    options,
  };
};
express.static = function staticMiddleware(root) {
  return {
    __xgnosisBuiltin: 'express.static',
    root,
  };
};
`;

  if (moduleFlavor === 'cjs') {
    return `${source}
module.exports = express;
module.exports.default = express;
module.exports.Router = express.Router;
module.exports.json = express.json;
module.exports.urlencoded = express.urlencoded;
module.exports.static = express.static;
`;
  }

  return `${source}
export default express;
export const Router = express.Router;
export const json = express.json;
export const urlencoded = express.urlencoded;
const expressStatic = express.static;
export { expressStatic as static };
`;
}

function renderHonoShim(moduleFlavor) {
  const source = `
const REGISTRY_KEY = ${JSON.stringify(REGISTRY_KEY)};
function registry() {
  if (!Array.isArray(globalThis[REGISTRY_KEY])) {
    globalThis[REGISTRY_KEY] = [];
  }
  return globalThis[REGISTRY_KEY];
}
function normalizePathname(value) {
  if (!value || value === '*') return '/';
  const path = String(value).split('?', 1)[0] || '/';
  return path.startsWith('/') ? path : \`/\${path}\`;
}
function trimTrailingSlash(value) {
  return value.length > 1 && value.endsWith('/') ? value.slice(0, -1) : value;
}
function joinMountedPath(prefix, childPath) {
  if (!prefix || prefix === '*' || prefix === '/*') return childPath;
  const normalizedPrefix = trimTrailingSlash(normalizePathname(prefix));
  const normalizedChild =
    childPath === '*' || childPath === '/*' ? '' : trimTrailingSlash(normalizePathname(childPath));
  if (!normalizedChild || normalizedChild === '/') return normalizedPrefix;
  return \`\${normalizedPrefix}\${normalizedChild}\`;
}
class Hono {
  constructor() {
    this.__xgnosis = {
      framework: 'hono',
      routes: [],
      middleware: [],
    };
    registry().push(this);
  }
  use(...args) {
    const [first, ...rest] = args;
    const hasPath = typeof first === 'string';
    const routePath = hasPath ? first : '*';
    const handlers = hasPath ? rest : args;
    this.__xgnosis.middleware.push({
      path: routePath,
      handlers,
    });
    return this;
  }
  route(prefix, child) {
    if (!child || !child.__xgnosis) {
      return this;
    }
    for (const route of child.__xgnosis.routes) {
      this.__xgnosis.routes.push({
        method: route.method,
        path: joinMountedPath(prefix, route.path),
        handlers: route.handlers,
      });
    }
    for (const middleware of child.__xgnosis.middleware) {
      this.__xgnosis.middleware.push({
        path: joinMountedPath(prefix, middleware.path),
        handlers: middleware.handlers,
      });
    }
    return this;
  }
}
for (const method of ['get', 'post', 'put', 'delete', 'patch', 'all']) {
  Hono.prototype[method] = function(routePath, ...handlers) {
    this.__xgnosis.routes.push({
      method: method.toUpperCase(),
      path: normalizePathname(routePath),
      handlers,
    });
    return this;
  };
}
`;

  if (moduleFlavor === 'cjs') {
    return `${source}
module.exports = {
  Hono,
  default: Hono,
};
`;
  }

  return `${source}
export { Hono };
export default Hono;
`;
}

function renderKoaShim(moduleFlavor) {
  const source = `
const REGISTRY_KEY = ${JSON.stringify(REGISTRY_KEY)};
function registry() {
  if (!Array.isArray(globalThis[REGISTRY_KEY])) {
    globalThis[REGISTRY_KEY] = [];
  }
  return globalThis[REGISTRY_KEY];
}
function normalizePathname(value) {
  if (!value || value === '*') return '/';
  const path = String(value).split('?', 1)[0] || '/';
  return path.startsWith('/') ? path : \`/\${path}\`;
}
function trimTrailingSlash(value) {
  return value.length > 1 && value.endsWith('/') ? value.slice(0, -1) : value;
}
function joinMountedPath(prefix, childPath) {
  if (!prefix || prefix === '*' || prefix === '/*') return childPath;
  const normalizedPrefix = trimTrailingSlash(normalizePathname(prefix));
  const normalizedChild =
    childPath === '*' || childPath === '/*' ? '' : trimTrailingSlash(normalizePathname(childPath));
  if (!normalizedChild || normalizedChild === '/') return normalizedPrefix;
  return \`\${normalizedPrefix}\${normalizedChild}\`;
}
class Koa {
  constructor() {
    this.__xgnosis = {
      framework: 'koa',
      listened: false,
      routes: [],
      middleware: [],
    };
    registry().push(this);
  }
  use(handler) {
    if (handler && handler.__xgnosisBuiltin === 'koa.router.routes') {
      const router = handler.router;
      for (const route of router.routes) {
        this.__xgnosis.routes.push({
          method: route.method,
          path: joinMountedPath(router.prefix, route.path),
          handlers: route.handlers,
        });
      }
      for (const middleware of router.middleware) {
        this.__xgnosis.middleware.push({
          path: joinMountedPath(router.prefix, middleware.path ?? '*'),
          handlers: middleware.handlers,
        });
      }
      return this;
    }
    if (handler && handler.__xgnosisBuiltin === 'koa.router.allowedMethods') {
      return this;
    }
    this.__xgnosis.middleware.push({
      path: '*',
      handlers: [handler],
    });
    return this;
  }
  listen() {
    this.__xgnosis.listened = true;
    return {
      close() {},
    };
  }
}
`;

  if (moduleFlavor === 'cjs') {
    return `${source}
module.exports = Koa;
module.exports.default = Koa;
`;
  }

  return `${source}
export default Koa;
`;
}

function renderKoaRouterShim(moduleFlavor) {
  const source = `
function normalizePathname(value) {
  if (!value || value === '*') return '/';
  const path = String(value).split('?', 1)[0] || '/';
  return path.startsWith('/') ? path : \`/\${path}\`;
}
class Router {
  constructor(options = {}) {
    this.__xgnosis = {
      framework: 'koa-router',
      prefix: options.prefix ? normalizePathname(options.prefix) : '/',
      routes: [],
      middleware: [],
    };
  }
  prefix(value) {
    this.__xgnosis.prefix = normalizePathname(value);
    return this;
  }
  use(...args) {
    const [first, ...rest] = args;
    const hasPath = typeof first === 'string';
    const routePath = hasPath ? normalizePathname(first) : '*';
    const handlers = hasPath ? rest : args;
    this.__xgnosis.middleware.push({
      path: routePath,
      handlers,
    });
    return this;
  }
  register(method, routePath, ...handlers) {
    this.__xgnosis.routes.push({
      method,
      path: normalizePathname(routePath),
      handlers,
    });
    return this;
  }
  routes() {
    return {
      __xgnosisBuiltin: 'koa.router.routes',
      router: this.__xgnosis,
    };
  }
  allowedMethods() {
    return {
      __xgnosisBuiltin: 'koa.router.allowedMethods',
      router: this.__xgnosis,
    };
  }
}
for (const method of ['get', 'post', 'put', 'delete', 'patch', 'all']) {
  Router.prototype[method] = function(routePath, ...handlers) {
    return this.register(method.toUpperCase(), routePath, ...handlers);
  };
}
`;

  if (moduleFlavor === 'cjs') {
    return `${source}
module.exports = Router;
module.exports.default = Router;
`;
  }

  return `${source}
export default Router;
`;
}

function renderFastifyShim(moduleFlavor) {
  const source = `
const REGISTRY_KEY = ${JSON.stringify(REGISTRY_KEY)};
function registry() {
  if (!Array.isArray(globalThis[REGISTRY_KEY])) {
    globalThis[REGISTRY_KEY] = [];
  }
  return globalThis[REGISTRY_KEY];
}
function normalizePathname(value) {
  if (!value || value === '*') return '/';
  const path = String(value).split('?', 1)[0] || '/';
  return path.startsWith('/') ? path : \`/\${path}\`;
}
function trimTrailingSlash(value) {
  return value.length > 1 && value.endsWith('/') ? value.slice(0, -1) : value;
}
function joinMountedPath(prefix, childPath) {
  if (!prefix || prefix === '*' || prefix === '/*') return childPath;
  const normalizedPrefix = trimTrailingSlash(normalizePathname(prefix));
  const normalizedChild =
    childPath === '*' || childPath === '/*' ? '' : trimTrailingSlash(normalizePathname(childPath));
  if (!normalizedChild || normalizedChild === '/') return normalizedPrefix;
  return \`\${normalizedPrefix}\${normalizedChild}\`;
}
function cloneHooks(record) {
  return {
    onRequest: [...record.hooks.onRequest],
    preHandler: [...record.hooks.preHandler],
  };
}
function createFastifyInstance(prefix = '/', registerInRegistry = true) {
  const record = {
    framework: 'fastify',
    listened: false,
    prefix: normalizePathname(prefix),
    routes: [],
    hooks: {
      onRequest: [],
      preHandler: [],
    },
  };
  const app = {
    __xgnosis: record,
    addHook(name, handler) {
      if (name === 'onRequest' || name === 'preHandler') {
        record.hooks[name].push(handler);
      }
      return app;
    },
    register(plugin, options = {}) {
      const child = createFastifyInstance(
        joinMountedPath(record.prefix, options.prefix ?? '/'),
        false
      );
      if (typeof plugin === 'function') {
        if (plugin.length >= 3) {
          plugin(child, options, () => {});
        } else {
          plugin(child, options);
        }
      }
      for (const route of child.__xgnosis.routes) {
        record.routes.push(route);
      }
      return app;
    },
    route(options) {
      const methods = Array.isArray(options.method) ? options.method : [options.method ?? 'GET'];
      const preHandlers = Array.isArray(options.preHandler)
        ? options.preHandler
        : options.preHandler
          ? [options.preHandler]
          : [];
      for (const method of methods) {
        record.routes.push({
          method: String(method).toUpperCase(),
          path: joinMountedPath(record.prefix, options.url ?? options.path ?? '/'),
          handler: options.handler,
          preHandlers: [...record.hooks.preHandler, ...preHandlers],
          onRequestHooks: [...record.hooks.onRequest],
        });
      }
      return app;
    },
    listen(options = {}, callback) {
      record.listened = true;
      if (typeof callback === 'function') {
        callback(null, '');
      }
      return Promise.resolve('');
    },
  };
  for (const method of ['get', 'post', 'put', 'delete', 'patch', 'head', 'options', 'all']) {
    app[method] = (routePath, optionsOrHandler, maybeHandler) => {
      let handler = maybeHandler;
      let options = {};
      if (typeof optionsOrHandler === 'function') {
        handler = optionsOrHandler;
      } else if (optionsOrHandler && typeof optionsOrHandler === 'object') {
        options = optionsOrHandler;
        handler = maybeHandler ?? optionsOrHandler.handler;
      }
      const preHandlers = Array.isArray(options.preHandler)
        ? options.preHandler
        : options.preHandler
          ? [options.preHandler]
          : [];
      record.routes.push({
        method: method.toUpperCase(),
        path: joinMountedPath(record.prefix, routePath),
        handler,
        preHandlers: [...record.hooks.preHandler, ...preHandlers],
        onRequestHooks: [...record.hooks.onRequest],
      });
      return app;
    };
  }
  if (registerInRegistry) {
    registry().push(app);
  }
  return app;
}
function fastify() {
  return createFastifyInstance('/');
}
`;

  if (moduleFlavor === 'cjs') {
    return `${source}
module.exports = fastify;
module.exports.default = fastify;
`;
  }

  return `${source}
export default fastify;
`;
}

async function importCapturedSource(filePath, frameworkId) {
  const absolutePath = path.resolve(filePath);
  const sourceDirectory = path.dirname(absolutePath);
  const sourceText = fs.readFileSync(absolutePath, 'utf8');
  const moduleFlavor = detectModuleFlavor(sourceText, absolutePath);
  const tempExtension = compatibilityTempExtension(absolutePath, moduleFlavor);
  const nonce = `${Date.now()}_${Math.random().toString(36).slice(2, 10)}`;
  const tempSourcePath = path.join(
    sourceDirectory,
    `.${path.basename(
      absolutePath,
      path.extname(absolutePath)
    )}.xgnosis.${nonce}${tempExtension}`
  );
  const shimExtension = moduleFlavor === 'cjs' ? '.cjs' : '.mjs';
  const shims = frameworkShims(frameworkId, moduleFlavor);
  let rewrittenSource = sourceText;
  const shimFilePaths = [];

  for (const [index, shim] of shims.entries()) {
    const packageToken = shim.packageName
      .replace(/[^a-zA-Z0-9]+/g, '_')
      .replace(/^_+|_+$/g, '');
    const shimFilePath = path.join(
      sourceDirectory,
      `.${path.basename(
        absolutePath,
        path.extname(absolutePath)
      )}.xgnosis.${nonce}.${index}.${packageToken}.shim${shimExtension}`
    );
    const shimSpecifier = relativeShimSpecifier(
      sourceDirectory,
      shimFilePath,
      moduleFlavor
    );
    rewrittenSource = rewritePackageImport(
      rewrittenSource,
      shim.packageName,
      shimSpecifier
    );
    fs.writeFileSync(shimFilePath, shim.source, 'utf8');
    shimFilePaths.push(shimFilePath);
  }

  fs.writeFileSync(tempSourcePath, rewrittenSource, 'utf8');

  clearRegistry();

  try {
    await import(`${pathToFileURL(tempSourcePath).href}?t=${Date.now()}`);
    const apps = registry();
    const selected =
      apps.find((candidate) => candidate?.__xgnosis?.listened) ??
      apps.find((candidate) => candidate?.__xgnosis?.kind === 'app') ??
      apps.find(
        (candidate) => candidate?.__xgnosis?.framework === frameworkId
      ) ??
      null;

    if (!selected) {
      throw new Error(
        `No ${frameworkId} compatibility app could be captured from ${filePath}.`
      );
    }

    return selected;
  } finally {
    clearRegistry();
    fs.rmSync(tempSourcePath, { force: true });
    for (const shimFilePath of shimFilePaths) {
      fs.rmSync(shimFilePath, { force: true });
    }
  }
}

/**
 * Handles the gnosis execute Node Framework Compatibility workflow.
 */
export async function executeNodeFrameworkCompatibility(request) {
  const compatibility = request.compatibility;
  if (!compatibility || compatibility.family !== 'node-router') {
    throw new Error('Missing node-router compatibility request.');
  }

  if (
    compatibility.framework !== 'express' &&
    compatibility.framework !== 'hono' &&
    compatibility.framework !== 'koa' &&
    compatibility.framework !== 'fastify'
  ) {
    return {
      matched: false,
    };
  }

  const app = await importCapturedSource(
    request.filePath,
    compatibility.framework
  );
  const invocation = {
    method: String(compatibility.request.method).toUpperCase(),
    path: normalizePathname(compatibility.request.path),
    query: compatibility.request.query ?? '',
    headers: normalizeResponseHeaders(compatibility.request.headers ?? {}),
    body: compatibility.request.body ?? null,
  };

  if (compatibility.framework === 'hono') {
    return dispatchHonoApp(app, invocation);
  }

  if (compatibility.framework === 'koa') {
    return dispatchKoaApp(app, invocation);
  }

  if (compatibility.framework === 'fastify') {
    return dispatchFastifyApp(app, invocation);
  }

  return dispatchExpressApp(app, invocation);
}
