// @a0n/.ignored_gnosis/src/esc/cache-observation.ts
/**
 * TypeScript mirror of `lean/Lean/ForkRaceFoldTheorems/CacheObservation.lean`.
 *
 * Same names in camelCase: marginal fold work on a single access, tag resolver,
 * and summed marginal work over a trace. Use with ESC to report ledger-aligned
 * observations on WASM (or other) cache paths.
 *
 * @see FORMAL_LEDGER.md § Cache Observation — Hits as Zero-Marginal-Fold Reads
 */

export const CACHE_OBSERVATION_LEAN_MODULE =
  'ForkRaceFoldTheorems.CacheObservation' as const;

/** Positive Nat witness: search / cold-fill harder than verify (Lean `PositiveFoldWork`). */
export interface PositiveFoldWork {
  readonly val: number;
}

/**
 * Handles the gnosis positive Fold Work workflow.
 */
export function positiveFoldWork(n: number): PositiveFoldWork {
  if (!Number.isInteger(n) || n < 1) {
    throw new RangeError(
      `positiveFoldWork: expected integer >= 1, got ${String(n)}`
    );
  }
  return { val: n };
}

/** Outcome of resolving one query against a cache line (Lean `CacheQueryResolution`). */
export type CacheQueryResolution =
  | { readonly kind: 'hit' }
  | { readonly kind: 'cold_miss'; readonly w: PositiveFoldWork }
  | { readonly kind: 'stale_miss'; readonly w: PositiveFoldWork };

/** Marginal Reidemeister / fold work for this access only (Lean `marginal_fold_work`). */
export function marginalFoldWork(q: CacheQueryResolution): number {
  switch (q.kind) {
    case 'hit':
      return 0;
    case 'cold_miss':
    case 'stale_miss':
      return q.w.val;
  }
}

/** Tag view on a line (Lean `CacheLineView`). */
export interface CacheLineView {
  readonly occupied: boolean;
  readonly keyMatches: boolean;
  readonly entryValid: boolean;
}

/** Tag-driven resolver (Lean `resolveQuery`). */
export function resolveQuery(
  line: CacheLineView,
  w: PositiveFoldWork
): CacheQueryResolution {
  if (line.occupied && line.keyMatches && line.entryValid) {
    return { kind: 'hit' };
  }
  if (line.occupied && line.keyMatches && !line.entryValid) {
    return { kind: 'stale_miss', w };
  }
  return { kind: 'cold_miss', w };
}

/** Sum marginal fold work over a finite access list (Lean `totalMarginalFoldWork`). */
export function totalMarginalFoldWork(
  qs: readonly CacheQueryResolution[]
): number {
  let sum = 0;
  for (const q of qs) {
    sum += marginalFoldWork(q);
  }
  return sum;
}

/**
 * Map ESC tier lookup + signature check into `CacheQueryResolution`.
 * - No row / wrong key → cold_miss
 * - Row + key + invalid coherence → stale_miss
 * - Row + key + valid → hit
 */
export function classifyEscCacheLookup(input: {
  readonly cached: unknown | null;
  readonly signatureMatches: boolean;
  /** Default true. False models phantom / invalidation (Lean stale path). */
  readonly entryValid?: boolean;
  readonly w: PositiveFoldWork;
}): CacheQueryResolution {
  const occupied = input.cached !== null;
  const keyMatches = input.signatureMatches;
  const entryValid = input.entryValid ?? true;
  return resolveQuery({ occupied, keyMatches, entryValid }, input.w);
}

/** Payload attached to WASM ESC responses for telemetry / ledger alignment. */
export interface WasmEscObservation {
  readonly formalRef: typeof CACHE_OBSERVATION_LEAN_MODULE;
  readonly resolution: CacheQueryResolution;
  readonly marginalFoldWork: number;
}

/**
 * Handles the gnosis wasm Esc Observation workflow.
 */
export function wasmEscObservation(
  resolution: CacheQueryResolution
): WasmEscObservation {
  return {
    formalRef: CACHE_OBSERVATION_LEAN_MODULE,
    resolution,
    marginalFoldWork: marginalFoldWork(resolution),
  };
}
