// @a0n/.ignored_gnosis/moonshine-tauri/scripts/moonshine-skymesh-bridge.mjs
#!/usr/bin/env node
import { createHash, randomUUID } from 'node:crypto';
import { mkdirSync, readFileSync, realpathSync, statSync, writeFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { spawn } from 'node:child_process';
import { fileURLToPath } from 'node:url';

const args = process.argv.slice(2);
const arg = (name, fallback) => {
  const i = args.indexOf(name);
  return i >= 0 && args[i + 1] && !args[i + 1].startsWith('--') ? args[i + 1] : fallback;
};

const appDir = dirname(fileURLToPath(import.meta.url));
const gnosisRoot = join(appDir, '..', '..');
const sidecarDir = join(appDir, '..', 'src-tauri', 'binaries');
const env = process.env;

const NODE = arg('--node', env.SKYMESH_NODE || `moonshine-${randomUUID().slice(0, 8)}`);
const MODEL = arg('--model', env.MOONSHINE_MODEL_ID || 'qwen2.5-0.5b-instruct-q4');
const FAT_BASE = arg('--fatstation', env.FATSTATION_URL || env.FAT_STATION_URL || 'http://127.0.0.1:8000');
const FAT_PORT = Number(new URL(FAT_BASE).port || '8000');
const FAT_BIN = arg(
  '--fat-bin',
  env.FAT_STATION_BIN || join(sidecarDir, 'fat-station')
);
const KNOT = arg(
  '--knot',
  env.MOONSHINE_LLM_KNOT || join(env.HOME || '.', '.edgework', 'models', 'qwen2.5-0.5b-instruct.knot')
);
const TOKENIZER = arg('--tokenizer', env.MOONSHINE_TOKENIZER_GGUF || env.TOKENIZER_GGUF_PATH || '');
const MONITOR_RELAYS = splitCsv(
  arg(
    '--monitor-relays',
    env.SKYMESH_MONITOR_RELAYS ||
      env.SKYMESH_RELAY ||
      'wss://skymesh.forkjoin.ai/monitor/bridge,ws://127.0.0.1:8787/monitor/bridge'
  )
);
const PROTOCOL69_BASES = splitCsv(
  arg(
    '--protocol69-bases',
    env.SKYMESH_PROTOCOL69_BASES ||
      env.SKYMESH_BASE_URL ||
      'https://skymesh.forkjoin.ai,http://127.0.0.1:8787'
  )
);
const MESHES = splitCsv(arg('--meshes', env.SKYMESH_MESHES || env.SKYMESH_MESH || 'skymesh-global,skymesh-local'));
const CLAIM_INTERVAL_MS = Number(arg('--claim-interval-ms', env.SKYMESH_PROTOCOL69_CLAIM_MS || '5000'));
const BRIDGE_TOKEN = arg('--token', env.SKYMESH_BRIDGE_TOKEN || '');
const STATUS_PATH = arg('--status-path', env.MOONSHINE_SKYMESH_STATUS || join(env.MONSTER_GUARD_HOME || join(env.HOME || '.', '.moonshine.d'), 'skymesh-status.json'));

const STOP_TOKEN_IDS = new Set([151643, 151645]);
const DEFAULT_MAX_NEW_TOKENS = 64;

const log = (...parts) => console.log(`[moonshine-skymesh ${NODE}]`, ...parts);

const status = {
  enabled: true,
  running: true,
  node: NODE,
  model: MODEL,
  monitor_relays: MONITOR_RELAYS,
  protocol69_bases: PROTOCOL69_BASES,
  meshes: MESHES,
  last_cache_key: null,
  filled_cache_keys: [],
  last_error: null,
  updated_at: new Date().toISOString(),
};

function writeStatus(patch = {}) {
  Object.assign(status, patch, { updated_at: new Date().toISOString() });
  try {
    mkdirSync(dirname(STATUS_PATH), { recursive: true });
    writeFileSync(STATUS_PATH, `${JSON.stringify(status, null, 2)}\n`);
  } catch {
    // Status is diagnostic only; never break mesh participation.
  }
}

function splitCsv(value) {
  return String(value || '')
    .split(',')
    .map((item) => item.trim())
    .filter(Boolean);
}

function fileExists(path) {
  return !!statSync(path, { throwIfNoEntry: false })?.isFile();
}

function knotLayerCount(path) {
  try {
    const buf = readFileSync(path);
    if (buf.subarray(0, 4).toString('utf8') !== 'KNOT') return 28;
    const metaLen = buf.readUInt32LE(6);
    const meta = JSON.parse(buf.subarray(10, 10 + metaLen).toString('utf8'));
    const layers = meta?.config?.num_layers;
    return Number.isFinite(layers) ? Number(layers) : 28;
  } catch {
    return 28;
  }
}

function resolveTokenizer() {
  if (TOKENIZER && fileExists(TOKENIZER)) return TOKENIZER;
  const modelDirs = [
    env.MOONSHINE_MODEL_DIR,
    env.EDGEWORK_MODEL_DIR,
    join(env.HOME || '.', '.edgework', 'models'),
  ].filter(Boolean);
  const candidates = modelDirs.flatMap((dir) => [
    join(dir, 'Qwen2.5-0.5B-Instruct-Q4_K_M.gguf'),
    join(dir, 'deepseek-r1-qwen3-8b-q4_k_m.gguf'),
  ]);
  return candidates.find(fileExists) || '';
}

async function sleep(ms) {
  await new Promise((resolve) => setTimeout(resolve, ms));
}

async function postJson(url, body, timeoutMs = 120000) {
  const ctrl = new AbortController();
  const timer = setTimeout(() => ctrl.abort(new Error(`timeout ${timeoutMs}ms`)), timeoutMs);
  try {
    const response = await fetch(url, {
      method: 'POST',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify(body),
      signal: ctrl.signal,
    });
    const text = await response.text();
    if (!response.ok) throw new Error(`HTTP ${response.status} ${url}: ${text.slice(0, 300)}`);
    const json = text ? JSON.parse(text) : {};
    return json;
  } finally {
    clearTimeout(timer);
  }
}

async function fatHealth() {
  try {
    const response = await fetch(new URL('/health', FAT_BASE), { signal: AbortSignal.timeout(1500) });
    return response.ok;
  } catch {
    return false;
  }
}

async function ensureFatStation() {
  if (await fatHealth()) return;
  if (!fileExists(FAT_BIN)) throw new Error(`fat-station binary not found: ${FAT_BIN}`);
  if (!fileExists(KNOT)) throw new Error(`model knot not found: ${KNOT}`);

  const tokenizer = resolveTokenizer();
  const childEnv = {
    ...env,
    ...(tokenizer ? { TOKENIZER_GGUF_PATH: tokenizer } : {}),
  };
  const layers = `0..${knotLayerCount(KNOT)}`;
  const child = spawn(FAT_BIN, ['--knot', KNOT, '--port', String(FAT_PORT), '--role', 'both', '--layers', layers], {
    env: childEnv,
    stdio: 'ignore',
    detached: true,
  });
  child.unref();
  log('spawned fat-station', FAT_BIN, '--knot', KNOT, '--layers', layers);

  for (let i = 0; i < 120; i += 1) {
    if (await fatHealth()) return;
    await sleep(1000);
  }
  throw new Error(`fat-station did not become healthy at ${FAT_BASE}`);
}

function answerPrompt(prompt) {
  return `Answer the user directly and concisely.\nUser: ${String(prompt || '').trim()}\nAnswer:`;
}

async function tokenize(text) {
  const body = await postJson(new URL('/tokenize', FAT_BASE), { text, add_bos: false }, 30000);
  const tokens = Array.isArray(body.tokens) ? body.tokens.filter(Number.isInteger) : [];
  if (!tokens.length) throw new Error('fat-station tokenize returned no tokens');
  return tokens;
}

async function decodeNext(token, position) {
  const response = await fetch(new URL('/decode-next', FAT_BASE), {
    method: 'POST',
    headers: { 'X-Token': String(token), 'X-Position': String(position) },
    body: new Uint8Array(),
    signal: AbortSignal.timeout(180000),
  });
  const text = await response.text();
  if (!response.ok) throw new Error(`/decode-next HTTP ${response.status}: ${text.slice(0, 300)}`);
  const json = text ? JSON.parse(text) : {};
  if (!Number.isInteger(json.token_id)) throw new Error('/decode-next missing token_id');
  return json.token_id;
}

async function detokenize(tokens) {
  if (!tokens.length) return '';
  const body = await postJson(new URL('/detokenize', FAT_BASE), { token_ids: tokens }, 30000);
  return cleanText(typeof body.text === 'string' ? body.text : '');
}

function cleanText(text) {
  return String(text)
    .split('<|im_end|>')[0]
    .split('|im_end|')[0]
    .split('<｜begin▁of▁sentence｜>')[0]
    .split('\nUser:')[0]
    .split('\nHuman:')[0]
    .trim();
}

async function generateFromTokens(promptTokens, maxNewTokens = DEFAULT_MAX_NEW_TOKENS) {
  const generated = [];
  let feedToken = promptTokens[0];
  let pos = 0;
  const promptLen = promptTokens.length;
  const totalSteps = promptLen + Math.max(0, maxNewTokens - 1);
  for (let step = 0; step < totalSteps; step += 1) {
    const next = await decodeNext(feedToken, pos);
    const isGenerationStep = pos + 1 >= promptLen;
    if (isGenerationStep) {
      if (STOP_TOKEN_IDS.has(next)) break;
      generated.push(next);
      if (generated.length >= maxNewTokens) break;
    }
    pos += 1;
    feedToken = pos < promptLen ? promptTokens[pos] : next;
  }
  return generated;
}

async function infer(prompt, maxNewTokens) {
  await ensureFatStation();
  const promptTokens = await tokenize(answerPrompt(prompt));
  const outTokens = await generateFromTokens(promptTokens, maxNewTokens || DEFAULT_MAX_NEW_TOKENS);
  return { text: await detokenize(outTokens), promptTokens, outTokens };
}

async function inferProtocol69Work(work) {
  const maxNewTokens = Number(work.maxNewTokens || work?.leg?.operation?.maxNewTokens || DEFAULT_MAX_NEW_TOKENS);
  await ensureFatStation();
  const promptTokens = Array.isArray(work.tokens) && work.tokens.length
    ? work.tokens.filter(Number.isInteger)
    : await tokenize(answerPrompt(work.prompt || ''));
  const outTokens = await generateFromTokens(promptTokens, maxNewTokens);
  return { text: await detokenize(outTokens), promptTokens, outTokens };
}

function sha256(value) {
  return `sha256:${createHash('sha256').update(value).digest('hex')}`;
}

function isOptionalLocalBase(base) {
  try {
    const url = new URL(base);
    return url.hostname === '127.0.0.1' || url.hostname === 'localhost' || url.hostname === '::1';
  } catch {
    return false;
  }
}

function connectMonitor(relay, mesh) {
  if (typeof WebSocket === 'undefined') {
    log('Node 22+ WebSocket unavailable; monitor bridge disabled');
    return;
  }
  const sep = relay.includes('?') ? '&' : '?';
  const url = `${relay}${sep}mesh=${encodeURIComponent(mesh)}&node=${encodeURIComponent(`${NODE}-${mesh}`)}${BRIDGE_TOKEN ? `&token=${encodeURIComponent(BRIDGE_TOKEN)}` : ''}`;
  log('monitor connect', url);
  const ws = new WebSocket(url);
  let admitted = false;
  const hello = () => {
    if (ws.readyState === WebSocket.OPEN) {
      ws.send(JSON.stringify({ type: 'hello', nodeId: `${NODE}-${mesh}`, role: 'fat-client', models: [MODEL], ...(admitted ? { admitted: true } : {}) }));
    }
  };
  ws.addEventListener('open', hello);
  ws.addEventListener('message', async (event) => {
    let msg;
    try {
      msg = JSON.parse(typeof event.data === 'string' ? event.data : event.data.toString());
    } catch {
      return;
    }
    if (msg.type === 'preflight') {
      try {
        const result = await infer('The capital of France is', 8);
        const pass = /paris/i.test(result.text);
        ws.send(JSON.stringify({ type: 'preflight-result', id: msg.id, pass, token: 0, text: result.text.slice(0, 48), logit: 0, elapsedMs: 0, model: MODEL }));
        if (pass && !admitted) {
          admitted = true;
          hello();
        }
      } catch (error) {
        ws.send(JSON.stringify({ type: 'preflight-result', id: msg.id, pass: false, token: 0, text: '', logit: 0, elapsedMs: 0, model: MODEL, detail: String(error) }));
      }
      return;
    }
    if (msg.type === 'query') {
      const started = Date.now();
      try {
        const result = await infer(String(msg.prompt || ''), Number(msg.maxTokens) || DEFAULT_MAX_NEW_TOKENS);
        ws.send(JSON.stringify({ type: 'query-result', id: msg.id, reply: result.text, promptTokens: result.promptTokens, outTokens: result.outTokens, elapsedMs: Date.now() - started, model: MODEL }));
      } catch (error) {
        ws.send(JSON.stringify({ type: 'query-result', id: msg.id, reply: '', promptTokens: [], outTokens: [], elapsedMs: Date.now() - started, model: MODEL, detail: String(error) }));
      }
    }
  });
  ws.addEventListener('close', () => setTimeout(() => connectMonitor(relay, mesh), 3000));
  ws.addEventListener('error', () => {
    try {
      ws.close();
    } catch {
      // ignore
    }
  });
}

async function claimOnce(base) {
  const claim = await postJson(new URL('/protocol69/work/claim', base), {
    agentId: NODE,
    executor: 'nativeGen',
    skills: ['runGenerate'],
  }, 30000);
  if (claim.status !== 'claimed' || !claim.work) return false;
  const work = claim.work;
  writeStatus({ last_cache_key: work.cacheKey ?? null, last_error: null });
  const result = await inferProtocol69Work(work);
  const text = cleanText(result.text);
  const resultHash = sha256(JSON.stringify({ text, tokens: result.outTokens, cacheKey: work.cacheKey }));
  const ack = {
    ackKind: 'legAck',
    legAck: {
      schemaVersion: 'gnosis.protocol69.workflow-leg-ack.v1',
      contractId: 'gnosis.protocol69.universe-cache.workflow-leg-ack.v1',
      workflowId: work.workflowId,
      dispatchId: work.dispatchId,
      legId: work.legId,
      cacheKey: work.cacheKey,
      executor: work.executor || 'nativeGen',
      workKind: work.workKind || 'runGenerate',
      resultHash,
      answerText: text,
      provenance: 'moonshine:protocol69-fat-client',
      certificate: work.certificate,
      projection: { key: work.key, text, tokens: result.outTokens },
    },
  };
  const accepted = await postJson(new URL('/protocol69/work/ack', base), ack, 30000);
  const filledCacheKeys = [work.cacheKey, ...status.filled_cache_keys.filter((key) => key !== work.cacheKey)].slice(0, 10);
  writeStatus({ last_cache_key: work.cacheKey ?? null, filled_cache_keys: filledCacheKeys, last_error: accepted.ok ? null : JSON.stringify(accepted) });
  log('protocol69 ack', base, work.cacheKey, accepted.ok ? 'accepted' : JSON.stringify(accepted));
  return true;
}

async function claimLoop(base) {
  log('protocol69 claim loop', base);
  for (;;) {
    try {
      const claimed = await claimOnce(base);
      await sleep(claimed ? 250 : CLAIM_INTERVAL_MS);
    } catch (error) {
      if (!isOptionalLocalBase(base)) {
        writeStatus({ last_error: String(error) });
      }
      log('protocol69 claim error', base, String(error));
      await sleep(CLAIM_INTERVAL_MS);
    }
  }
}

function isMainModule() {
  try {
    return !!process.argv[1] && realpathSync(process.argv[1]) === realpathSync(fileURLToPath(import.meta.url));
  } catch {
    return false;
  }
}

function startSkymeshBridge() {
  for (const relay of MONITOR_RELAYS) {
    for (const mesh of MESHES) connectMonitor(relay, mesh);
  }
  for (const base of PROTOCOL69_BASES) {
    claimLoop(base.endsWith('/') ? base : `${base}/`);
  }

  writeStatus();
  log('started', { model: MODEL, fat: FAT_BASE, knot: KNOT, monitorRelays: MONITOR_RELAYS, protocol69Bases: PROTOCOL69_BASES });
}

if (isMainModule()) {
  startSkymeshBridge();
}

export {
  answerPrompt,
  cleanText,
  inferProtocol69Work,
  isOptionalLocalBase,
  knotLayerCount,
  postJson,
  sha256,
  splitCsv,
  startSkymeshBridge,
};
