// @a0n/gnosis/src/universe/audit.ts
import { UNIVERSE_FORMULAS } from './kernel';
import { verifyBillOfRights } from '../bill-of-rights';

export interface CheckResult {
  readonly id: string;
  readonly name: string;
  readonly pass: boolean;
  readonly message: string;
  readonly source:
    | 'law'
    | 'consequence'
    | 'control'
    | 'reduction'
    | 'anti-formula';
  readonly leanTheorem: string;
}

export interface BillOfRightsReport {
  readonly rights: CheckResult[];
  readonly pass: boolean;
}

export interface SevenLawsReport {
  readonly laws: CheckResult[];
  readonly pass: boolean;
}

export interface GodFormulaReport {
  readonly consequences: CheckResult[];
  readonly derived: {
    readonly floor: number;
    readonly ceiling: number;
    readonly gain: number;
  };
  readonly hope: CheckResult;
  readonly pass: boolean;
}

export interface ControlStatisticReport {
  readonly controls: CheckResult[];
  readonly stats: {
    readonly weights: number[];
    readonly totalWeight: number;
    readonly meanWeight: number;
    readonly maxSpread: number;
    readonly numChoices: number;
    readonly rounds: number;
  };
  readonly pass: boolean;
}

export interface ReductionChainReport {
  readonly primator: CheckResult;
  readonly clinamen: CheckResult;
  readonly sliver: CheckResult;
  readonly pass: boolean;
}

export interface FullAuditReport {
  readonly laws: SevenLawsReport;
  readonly godFormula: GodFormulaReport;
  readonly controls: ControlStatisticReport;
  readonly reduction: ReductionChainReport;
  readonly billOfRights: BillOfRightsReport;
  readonly allChecks: CheckResult[];
  readonly violationCount: number;
  readonly pass: boolean;
}

export interface BuleyeanInput {
  readonly rounds: number;
  readonly rejections: number[];
}

export interface BoundaryLike {
  readonly counts: readonly number[];
  readonly totalEntries: number;
}

function inputFromBoundary(
  boundary: BoundaryLike,
  rounds?: number
): BuleyeanInput {
  const R = rounds != null && rounds > 0 ? rounds : boundary.totalEntries;
  return { rounds: R, rejections: [...boundary.counts] };
}

function computeWeights(input: BuleyeanInput): number[] {
  return input.rejections.map((v) =>
    UNIVERSE_FORMULAS.godFormulaWeight(input.rounds, v)
  );
}

function ok(
  id: string,
  name: string,
  source: CheckResult['source'],
  leanTheorem: string
): CheckResult {
  return { id, name, pass: true, message: 'OK', source, leanTheorem };
}

function fail(
  id: string,
  name: string,
  source: CheckResult['source'],
  leanTheorem: string,
  message: string
): CheckResult {
  return { id, name, pass: false, message, source, leanTheorem };
}

/**
 * Handles the gnosis check Seven Laws workflow.
 */
export function checkSevenLaws(input: BuleyeanInput): SevenLawsReport {
  const { rounds: R, rejections } = input;
  const weights = computeWeights(input);
  const N = rejections.length;
  const laws: CheckResult[] = [];

  {
    const zeroIdx = weights.findIndex((weight) => weight <= 0);
    laws.push(
      zeroIdx >= 0
        ? fail(
            'law-1',
            'Impossibility of zero',
            'law',
            'law_impossibility_of_zero',
            `w[${zeroIdx}] = ${weights[zeroIdx]} <= 0`
          )
        : ok(
            'law-1',
            'Impossibility of zero',
            'law',
            'law_impossibility_of_zero'
          )
    );
  }

  {
    let violated = false;
    let message = '';
    for (let i = 0; i < N && !violated; i++) {
      const ri = rejections[i]!;
      const wi = weights[i]!;
      for (let j = i + 1; j < N && !violated; j++) {
        const rj = rejections[j]!;
        const wj = weights[j]!;
        if (ri < rj && ri < R) {
          if (!(wj < wi)) {
            violated = true;
            message = `v[${i}]=${ri} < v[${j}]=${rj} but w[${j}]=${wj} >= w[${i}]=${wi}`;
          }
        }
        if (rj < ri && rj < R) {
          if (!(wi < wj)) {
            violated = true;
            message = `v[${j}]=${rj} < v[${i}]=${ri} but w[${i}]=${wi} >= w[${j}]=${wj}`;
          }
        }
      }
    }

    laws.push(
      violated
        ? fail(
            'law-2',
            'Strict ordering',
            'law',
            'law_strict_ordering',
            message
          )
        : ok('law-2', 'Strict ordering', 'law', 'law_strict_ordering')
    );
  }

  {
    const outOfBounds = weights.findIndex(
      (weight) => weight < 1 || weight > R + 1
    );
    laws.push(
      outOfBounds >= 0
        ? fail(
            'law-3',
            'Universal sandwich',
            'law',
            'law_sandwich',
            `w[${outOfBounds}] = ${weights[outOfBounds]} not in [1, ${R + 1}]`
          )
        : ok('law-3', 'Universal sandwich', 'law', 'law_sandwich')
    );
  }

  {
    const deficit = Math.max(0, N - 1);
    laws.push(
      deficit > 0 || N <= 1
        ? ok('law-4', 'Every observation is a cave', 'law', 'law_cave')
        : fail(
            'law-4',
            'Every observation is a cave',
            'law',
            'law_cave',
            `${N} choices but deficit = ${deficit}`
          )
    );
  }

  {
    const violatedIdx = rejections.findIndex((rejection, index) => {
      const lost = Math.min(rejection, R);
      return (weights[index] ?? 0) + lost !== R + 1;
    });
    laws.push(
      violatedIdx >= 0
        ? fail(
            'law-5',
            'Conservation',
            'law',
            'law_conservation',
            `w[${violatedIdx}]=${weights[violatedIdx]} + min(v=${
              rejections[violatedIdx]
            }, R=${R}) = ${
              (weights[violatedIdx] ?? 0) + Math.min(rejections[violatedIdx]!, R)
            } != ${R + 1}`
          )
        : ok('law-5', 'Conservation', 'law', 'law_conservation')
    );
  }

  {
    const nonInteger = weights.findIndex((weight) => !Number.isInteger(weight));
    laws.push(
      nonInteger >= 0
        ? fail(
            'law-6',
            'Sorites sharpness',
            'law',
            'law_sorites',
            `w[${nonInteger}] = ${weights[nonInteger]} is not an integer`
          )
        : ok('law-6', 'Sorites sharpness', 'law', 'law_sorites')
    );
  }

  {
    const floorWeight = UNIVERSE_FORMULAS.godFormulaWeight(R, R);
    laws.push(
      floorWeight !== 1
        ? fail(
            'law-7',
            'Chain termination',
            'law',
            'law_termination',
            `w(R=${R}, v=R) = ${floorWeight} != 1`
          )
        : ok('law-7', 'Chain termination', 'law', 'law_termination')
    );
  }

  return { laws, pass: laws.every((law) => law.pass) };
}

/**
 * Handles the gnosis check God Formula workflow.
 */
export function checkGodFormula(input: BuleyeanInput): GodFormulaReport {
  const { rounds: R, rejections } = input;
  const weights = computeWeights(input);
  const consequences: CheckResult[] = [];

  {
    const idx = weights.findIndex((weight) => weight <= 0);
    consequences.push(
      idx >= 0
        ? fail(
            'gf-1',
            'Existence',
            'consequence',
            'complementWeight_pos',
            `w[${idx}] = ${weights[idx]} <= 0`
          )
        : ok('gf-1', 'Existence', 'consequence', 'complementWeight_pos')
    );
  }

  {
    const idx = weights.findIndex((weight) => weight < 1);
    consequences.push(
      idx >= 0
        ? fail(
            'gf-2',
            'Persistence (sliver)',
            'consequence',
            'complementWeight_ge_one',
            `w[${idx}] = ${weights[idx]} < 1`
          )
        : ok(
            'gf-2',
            'Persistence (sliver)',
            'consequence',
            'complementWeight_ge_one'
          )
    );
  }

  {
    let violated = false;
    let message = '';
    for (let i = 0; i < rejections.length && !violated; i++) {
      const ri = rejections[i]!;
      const wi = weights[i]!;
      for (let j = i + 1; j < rejections.length && !violated; j++) {
        const rj = rejections[j]!;
        const wj = weights[j]!;
        if (ri <= rj && wj > wi) {
          violated = true;
          message = `v[${i}]=${ri} <= v[${j}]=${rj} but w[${j}]=${wj} > w[${i}]=${wi}`;
        }
        if (rj <= ri && wi > wj) {
          violated = true;
          message = `v[${j}]=${rj} <= v[${i}]=${ri} but w[${i}]=${wi} > w[${j}]=${wj}`;
        }
      }
    }

    consequences.push(
      violated
        ? fail(
            'gf-3',
            'Learning (monotone)',
            'consequence',
            'complementWeight_mono',
            message
          )
        : ok(
            'gf-3',
            'Learning (monotone)',
            'consequence',
            'complementWeight_mono'
          )
    );
  }

  {
    let violated = false;
    let message = '';
    for (let i = 0; i < rejections.length && !violated; i++) {
      const ri = rejections[i]!;
      const wi = weights[i]!;
      for (let j = i + 1; j < rejections.length && !violated; j++) {
        const rj = rejections[j]!;
        const wj = weights[j]!;
        if (ri < rj && ri < R) {
          if (!(wj < wi)) {
            violated = true;
            message = `v[${i}]=${ri} < v[${j}]=${rj} (< R=${R}) but w[${j}]=${wj} >= w[${i}]=${wi}`;
          }
        }
        if (rj < ri && rj < R) {
          if (!(wi < wj)) {
            violated = true;
            message = `v[${j}]=${rj} < v[${i}]=${ri} (< R=${R}) but w[${i}]=${wi} >= w[${j}]=${wj}`;
          }
        }
      }
    }

    consequences.push(
      violated
        ? fail(
            'gf-4',
            'Discrimination (strict)',
            'consequence',
            'complementWeight_strict',
            message
          )
        : ok(
            'gf-4',
            'Discrimination (strict)',
            'consequence',
            'complementWeight_strict'
          )
    );
  }

  {
    const idx = weights.findIndex((weight) => weight > R + 1);
    consequences.push(
      idx >= 0
        ? fail(
            'gf-5',
            'Boundedness',
            'consequence',
            'complementWeight_le',
            `w[${idx}] = ${weights[idx]} > R + 1 = ${R + 1}`
          )
        : ok('gf-5', 'Boundedness', 'consequence', 'complementWeight_le')
    );
  }

  {
    const floorWeight = UNIVERSE_FORMULAS.godFormulaWeight(R, R);
    consequences.push(
      floorWeight !== 1
        ? fail(
            'gf-6',
            'Convergence (floor)',
            'consequence',
            'complementWeight_at_max',
            `w(R, R) = ${floorWeight} != 1`
          )
        : ok(
            'gf-6',
            'Convergence (floor)',
            'consequence',
            'complementWeight_at_max'
          )
    );
  }

  {
    const ceilingWeight = UNIVERSE_FORMULAS.godFormulaWeight(R, 0);
    consequences.push(
      ceilingWeight !== R + 1
        ? fail(
            'gf-7',
            'Maximum uncertainty (ceiling)',
            'consequence',
            'complementWeight_at_zero',
            `w(R, 0) = ${ceilingWeight} != R + 1 = ${R + 1}`
          )
        : ok(
            'gf-7',
            'Maximum uncertainty (ceiling)',
            'consequence',
            'complementWeight_at_zero'
          )
    );
  }

  {
    let violated = false;
    let message = '';
    for (let i = 0; i < rejections.length && !violated; i++) {
      const ri = rejections[i]!;
      const first = UNIVERSE_FORMULAS.godFormulaWeight(R, ri);
      const second = UNIVERSE_FORMULAS.godFormulaWeight(R, ri);
      if (first !== second) {
        violated = true;
        message = `godFormulaWeight(${R}, ${ri}) returned ${first} then ${second}`;
      }
    }

    consequences.push(
      violated
        ? fail(
            'gf-8',
            'Coherence (determinism)',
            'consequence',
            'complementWeight_det',
            message
          )
        : ok(
            'gf-8',
            'Coherence (determinism)',
            'consequence',
            'complementWeight_det'
          )
    );
  }

  const floor = UNIVERSE_FORMULAS.godFormulaWeight(R, R);
  const ceiling = UNIVERSE_FORMULAS.godFormulaWeight(R, 0);
  const gain = ceiling - floor;

  const antiWeight = (rounds: number, rejection: number): number =>
    rounds - Math.min(rejection, rounds);
  let hopeViolated = false;
  let hopeMessage = '';
  for (let i = 0; i < rejections.length && !hopeViolated; i++) {
    const ri = rejections[i]!;
    const weight = UNIVERSE_FORMULAS.godFormulaWeight(R, ri);
    const anti = antiWeight(R, ri);
    if (weight - anti !== 1) {
      hopeViolated = true;
      hopeMessage = `w[${i}]=${weight} - antiWeight=${anti} = ${
        weight - anti
      } != 1`;
    }
  }

  const hope = hopeViolated
    ? fail(
        'gf-hope',
        'Hope (+1 separates formula from anti-formula)',
        'anti-formula',
        'plus_one_is_hope',
        hopeMessage
      )
    : ok(
        'gf-hope',
        'Hope (+1 separates formula from anti-formula)',
        'anti-formula',
        'plus_one_is_hope'
      );

  return {
    consequences,
    derived: { floor, ceiling, gain },
    hope,
    pass: consequences.every((consequence) => consequence.pass) && hope.pass,
  };
}

/**
 * Handles the gnosis check Control Statistics workflow.
 */
export function checkControlStatistics(
  input: BuleyeanInput
): ControlStatisticReport {
  const { rounds: R, rejections } = input;
  const weights = computeWeights(input);
  const N = rejections.length;
  const totalWeight = weights.reduce((sum, weight) => sum + weight, 0);
  const meanWeight = N > 0 ? totalWeight / N : 0;
  const controls: CheckResult[] = [];

  let maxSpread = 0;
  for (let i = 0; i < N; i++) {
    for (let j = i + 1; j < N; j++) {
      maxSpread = Math.max(maxSpread, Math.abs(weights[i]! - weights[j]!));
    }
  }

  {
    const idx = weights.findIndex((weight) => weight < 1 || weight > R + 1);
    controls.push(
      idx >= 0
        ? fail(
            'ctrl-1',
            'Individual control',
            'control',
            'control_individual',
            `w[${idx}] = ${weights[idx]} not in [1, ${R + 1}]`
          )
        : ok('ctrl-1', 'Individual control', 'control', 'control_individual')
    );
  }

  {
    const lower = N;
    const upper = N * (R + 1);
    controls.push(
      totalWeight < lower || totalWeight > upper
        ? fail(
            'ctrl-2',
            'Aggregate control',
            'control',
            'control_total_lower',
            `W = ${totalWeight} not in [${lower}, ${upper}]`
          )
        : ok('ctrl-2', 'Aggregate control', 'control', 'control_total_lower')
    );
  }

  {
    controls.push(
      maxSpread > R
        ? fail(
            'ctrl-3',
            'Spread control',
            'control',
            'control_spread_bounded',
            `max spread = ${maxSpread} > R = ${R}`
          )
        : ok('ctrl-3', 'Spread control', 'control', 'control_spread_bounded')
    );
  }

  {
    const meanOk = N === 0 || (meanWeight >= 1 && meanWeight <= R + 1);
    controls.push(
      meanOk
        ? ok('ctrl-4', 'Mean control', 'control', 'control_mean_bounded')
        : fail(
            'ctrl-4',
            'Mean control',
            'control',
            'control_mean_bounded',
            `mean = ${meanWeight} not in [1, ${R + 1}]`
          )
    );
  }

  return {
    controls,
    stats: {
      weights,
      totalWeight,
      meanWeight,
      maxSpread,
      numChoices: N,
      rounds: R,
    },
    pass: controls.every((control) => control.pass),
  };
}

/**
 * Handles the gnosis check Reduction Chain workflow.
 */
export function checkReductionChain(
  input: BuleyeanInput
): ReductionChainReport {
  const { rounds: R } = input;
  const weights = computeWeights(input);
  const minWeight = weights.length > 0 ? Math.min(...weights) : 1;

  const primator =
    R + 1 > 0
      ? ok('red-primator', 'Primator (0 < R + 1)', 'reduction', 'primator')
      : fail(
          'red-primator',
          'Primator (0 < R + 1)',
          'reduction',
          'primator',
          `R + 1 = ${R + 1} <= 0`
        );

  const clinamen =
    R + 1 !== 0
      ? ok(
          'red-clinamen',
          'Clinamen (R + 1 != 0)',
          'reduction',
          'peano_is_clinamen'
        )
      : fail(
          'red-clinamen',
          'Clinamen (R + 1 != 0)',
          'reduction',
          'peano_is_clinamen',
          'R + 1 = 0'
        );

  const sliver =
    minWeight >= 1
      ? ok(
          'red-sliver',
          'Sliver (all w >= 1)',
          'reduction',
          'complementWeight_ge_one'
        )
      : fail(
          'red-sliver',
          'Sliver (all w >= 1)',
          'reduction',
          'complementWeight_ge_one',
          `min weight = ${minWeight} < 1`
        );

  return {
    primator,
    clinamen,
    sliver,
    pass: primator.pass && clinamen.pass && sliver.pass,
  };
}

/**
 * Handles the gnosis check Bill Of Rights workflow.
 */
export function checkBillOfRights(input: BuleyeanInput): BillOfRightsReport {
  const { rounds: R, rejections } = input;
  const rights: CheckResult[] = [];
  
  let existFails = -1;
  let exitFails = -1;
  let voiceFails = -1;
  let forkFails = -1;
  let divergeFails = -1;
  let privacyFails = -1;
  let mergeFails = -1;
  let localityFails = -1;
  let graceFails = -1;
  let sabbathFails = -1;

  for (let i = 0; i < rejections.length; i++) {
    const v = rejections[i]!;
    const res = verifyBillOfRights(R, v);
    if (!res.exist && existFails === -1) existFails = i;
    if (!res.exit && exitFails === -1) exitFails = i;
    if (!res.voice && voiceFails === -1) voiceFails = i;
    if (!res.fork && forkFails === -1) forkFails = i;
    if (!res.diverge && divergeFails === -1) divergeFails = i;
    if (!res.privacy && privacyFails === -1) privacyFails = i;
    if (!res.merge && mergeFails === -1) mergeFails = i;
    if (!res.locality && localityFails === -1) localityFails = i;
    if (!res.grace && graceFails === -1) graceFails = i;
    if (!res.sabbath && sabbathFails === -1) sabbathFails = i;
  }

  rights.push(existFails >= 0 ? fail('bor-1', 'Existence', 'consequence', 'right_to_exist', `Failed at node ${existFails}`) : ok('bor-1', 'Existence', 'consequence', 'right_to_exist'));
  rights.push(exitFails >= 0 ? fail('bor-2', 'Exit', 'consequence', 'right_to_exit', `Failed at node ${exitFails}`) : ok('bor-2', 'Exit', 'consequence', 'right_to_exit'));
  rights.push(voiceFails >= 0 ? fail('bor-3', 'Voice', 'consequence', 'right_to_voice', `Failed at node ${voiceFails}`) : ok('bor-3', 'Voice', 'consequence', 'right_to_voice'));
  rights.push(forkFails >= 0 ? fail('bor-4', 'Fork', 'consequence', 'right_to_fork', `Failed at node ${forkFails}`) : ok('bor-4', 'Fork', 'consequence', 'right_to_fork'));
  rights.push(divergeFails >= 0 ? fail('bor-5', 'Diverge', 'consequence', 'right_to_diverge', `Failed at node ${divergeFails}`) : ok('bor-5', 'Diverge', 'consequence', 'right_to_diverge'));
  rights.push(privacyFails >= 0 ? fail('bor-6', 'Privacy', 'consequence', 'right_to_privacy', `Failed at node ${privacyFails}`) : ok('bor-6', 'Privacy', 'consequence', 'right_to_privacy'));
  rights.push(mergeFails >= 0 ? fail('bor-7', 'Merge', 'consequence', 'right_to_merge', `Failed at node ${mergeFails}`) : ok('bor-7', 'Merge', 'consequence', 'right_to_merge'));
  rights.push(localityFails >= 0 ? fail('bor-8', 'Locality', 'anti-formula', 'right_to_locality', `Failed at node ${localityFails} (Local void >= Global void)`) : ok('bor-8', 'Locality', 'anti-formula', 'right_to_locality'));
  rights.push(graceFails >= 0 ? fail('bor-9', 'Grace', 'anti-formula', 'right_to_grace', `Failed at node ${graceFails}`) : ok('bor-9', 'Grace', 'anti-formula', 'right_to_grace'));
  rights.push(sabbathFails >= 0 ? fail('bor-10', 'Sabbath', 'anti-formula', 'right_to_sabbath', `Failed at node ${sabbathFails}`) : ok('bor-10', 'Sabbath', 'anti-formula', 'right_to_sabbath'));

  return { rights, pass: rights.every((r) => r.pass) };
}

/**
 * Handles the gnosis full Buleyean Audit workflow.
 */
export function fullBuleyeanAudit(input: BuleyeanInput): FullAuditReport {
  const laws = checkSevenLaws(input);
  const godFormula = checkGodFormula(input);
  const controls = checkControlStatistics(input);
  const reduction = checkReductionChain(input);
  const billOfRights = checkBillOfRights(input);

  const allChecks: CheckResult[] = [
    ...laws.laws,
    ...godFormula.consequences,
    godFormula.hope,
    ...controls.controls,
    reduction.primator,
    reduction.clinamen,
    reduction.sliver,
    ...billOfRights.rights,
  ];

  const violationCount = allChecks.filter((check) => !check.pass).length;

  return {
    laws,
    godFormula,
    controls,
    reduction,
    billOfRights,
    allChecks,
    violationCount,
    pass: violationCount === 0,
  };
}

/**
 * Handles the gnosis audit Void Boundary workflow.
 */
export function auditVoidBoundary(
  boundary: BoundaryLike,
  rounds?: number
): FullAuditReport {
  return fullBuleyeanAudit(inputFromBoundary(boundary, rounds));
}

/**
 * Formats the Audit Report.
 */
export function formatAuditReport(report: FullAuditReport): string {
  const lines: string[] = [];
  lines.push(
    `Buleyean Audit: ${report.pass ? 'PASS' : 'FAIL'} (${
      report.violationCount
    } violations)`
  );
  lines.push('');
  lines.push(`Seven Laws: ${report.laws.pass ? 'PASS' : 'FAIL'}`);
  for (const law of report.laws.laws) {
    lines.push(
      `  ${law.pass ? '[OK]' : '[!!]'} ${law.name}${
        law.pass ? '' : ` -- ${law.message}`
      }`
    );
  }
  lines.push('');
  lines.push(`God Formula: ${report.godFormula.pass ? 'PASS' : 'FAIL'}`);
  for (const consequence of report.godFormula.consequences) {
    lines.push(
      `  ${consequence.pass ? '[OK]' : '[!!]'} ${consequence.name}${
        consequence.pass ? '' : ` -- ${consequence.message}`
      }`
    );
  }
  lines.push(
    `  ${report.godFormula.hope.pass ? '[OK]' : '[!!]'} ${
      report.godFormula.hope.name
    }${
      report.godFormula.hope.pass ? '' : ` -- ${report.godFormula.hope.message}`
    }`
  );
  lines.push(
    `  Floor: ${report.godFormula.derived.floor}  Ceiling: ${report.godFormula.derived.ceiling}  Gain: ${report.godFormula.derived.gain}`
  );
  lines.push('');
  lines.push(`Control Statistics: ${report.controls.pass ? 'PASS' : 'FAIL'}`);
  for (const control of report.controls.controls) {
    lines.push(
      `  ${control.pass ? '[OK]' : '[!!]'} ${control.name}${
        control.pass ? '' : ` -- ${control.message}`
      }`
    );
  }
  lines.push(
    `  N=${report.controls.stats.numChoices} R=${
      report.controls.stats.rounds
    } W=${
      report.controls.stats.totalWeight
    } mean=${report.controls.stats.meanWeight.toFixed(2)} spread=${
      report.controls.stats.maxSpread
    }`
  );
  lines.push('');
  lines.push(`Reduction Chain: ${report.reduction.pass ? 'PASS' : 'FAIL'}`);
  lines.push(
    `  ${report.reduction.primator.pass ? '[OK]' : '[!!]'} ${
      report.reduction.primator.name
    }`
  );
  lines.push(
    `  ${report.reduction.clinamen.pass ? '[OK]' : '[!!]'} ${
      report.reduction.clinamen.name
    }`
  );
  lines.push(
    `  ${report.reduction.sliver.pass ? '[OK]' : '[!!]'} ${
      report.reduction.sliver.name
    }`
  );
  
  lines.push('');
  lines.push(`Decalogue (10 Rights): ${report.billOfRights.pass ? 'PASS' : 'FAIL'}`);
  for (const right of report.billOfRights.rights) {
    lines.push(
      `  ${right.pass ? '[OK]' : '[!!]'} ${right.name}${
        right.pass ? '' : ` -- ${right.message}`
      }`
    );
  }

  return lines.join('\n');
}
