// @a0n/gnosis/src/universe/certificate.ts
import {
  computeLawProgramMetrics,
  createLawState,
  CURRENT_DIMENSION,
  digestLawProgram,
  LAW_KERNEL_ID,
  APPLICATION_DIMENSION,
  sha256Hex,
  stableStringify,
  type LawProgram,
  type UniverseCertificate,
  type UniverseRuntimeMetadata,
} from './kernel';
import type { ClockworkVerdict, LaplaceWitness } from './verdict';

export interface ClockworkCertificate {
  readonly version: 1;
  readonly subject: string;
  readonly status: 'verified' | 'open' | 'violated';
  readonly voidBoundary: {
    readonly counts: readonly number[];
    readonly totalEntries: number;
    readonly teleportDeficit: number;
  };
  readonly deficit: {
    readonly total: number;
    readonly max: number;
    readonly openNodes: number;
    readonly totalNodes: number;
  };
  readonly conservation: {
    readonly vFork: number;
    readonly wFold: number;
    readonly qVent: number;
    readonly conserved: boolean;
  };
  readonly heat: {
    readonly totalHeat: number;
    readonly mergePoints: number;
  };
  readonly checks: readonly {
    readonly id: string;
    readonly pass: boolean;
    readonly theorem: string;
  }[];
  readonly verifiedAt: number;
  readonly sufficientStatisticTheorem: 'THM-VOID-BOUNDARY-SUFFICIENT-STATISTIC';
  readonly laplace?: LaplaceWitness;
}

/**
 * Handles the gnosis generate Certificate workflow.
 */
export function generateCertificate(
  subject: string,
  verdict: ClockworkVerdict
): ClockworkCertificate {
  return {
    version: 1,
    subject,
    status: verdict.status,
    voidBoundary: {
      counts: verdict.void.counts,
      totalEntries: verdict.void.totalEntries,
      teleportDeficit: verdict.void.teleportDeficit,
    },
    deficit: {
      total: verdict.deficit.totalDeficit,
      max: verdict.deficit.maxDeficit,
      openNodes: verdict.deficit.openNodes,
      totalNodes: verdict.deficit.totalNodes,
    },
    conservation: {
      vFork: verdict.conservation.vFork,
      wFold: verdict.conservation.wFold,
      qVent: verdict.conservation.qVent,
      conserved: verdict.conservation.conserved,
    },
    heat: {
      totalHeat: verdict.heat.totalHeat,
      mergePoints: verdict.heat.mergePoints,
    },
    checks: verdict.checks.map((check) => ({
      id: check.id,
      pass: check.pass,
      theorem: check.theorem,
    })),
    verifiedAt: verdict.verifiedAt,
    sufficientStatisticTheorem: 'THM-VOID-BOUNDARY-SUFFICIENT-STATISTIC',
    ...(verdict.laplace !== undefined && { laplace: verdict.laplace }),
  };
}

/**
 * Handles the gnosis generate Universe Certificate workflow.
 */
export function generateUniverseCertificate(
  program: LawProgram,
  verdict: ClockworkVerdict,
  clockworkCertificate: ClockworkCertificate = generateCertificate(
    program.subject ?? 'topology',
    verdict
  )
): UniverseCertificate {
  const metrics = computeLawProgramMetrics(program);
  const theoremRefs = [
    ...new Set(verdict.checks.map((check) => check.theorem).filter(Boolean)),
  ].sort();
  const sourceDigest = program.sourceDigest ?? digestLawProgram(program);
  const baseCertificate = {
    version: 1 as const,
    kernelId: LAW_KERNEL_ID,
    subject: program.subject ?? clockworkCertificate.subject,
    status: verdict.status,
    theoremRefs,
    capabilitySummary: program.capabilitySummary ?? {},
    dependencyCertificateDigests: [
      ...(program.dependencyCertificateDigests ?? []),
    ].sort(),
    sourceDigest,
    buildMetadata: {
      verifiedAt: verdict.verifiedAt,
      nodeCount: metrics.nodeCount,
      edgeCount: metrics.edgeCount,
      dimensionCount: metrics.dimensionCount,
      applicationDimension: APPLICATION_DIMENSION,
      currentDimension: CURRENT_DIMENSION,
      dimensionlessApplications: true as const,
      quantumPrimitivesExplicit: true as const,
    },
    dimension: APPLICATION_DIMENSION,
    currentDimension: CURRENT_DIMENSION,
    clockwork: clockworkCertificate,
  };

  return {
    ...baseCertificate,
    digest: sha256Hex(stableStringify(baseCertificate)),
  };
}

/**
 * Handles the gnosis attach Universe Runtime Metadata workflow.
 */
export function attachUniverseRuntimeMetadata(
  certificate: UniverseCertificate,
  runtimeMetadata: UniverseRuntimeMetadata
): UniverseCertificate {
  const { digest: _ignoredDigest, ...withoutDigest } = certificate;
  const baseCertificate = {
    ...withoutDigest,
    runtimeMetadata,
  };

  return {
    ...baseCertificate,
    digest: sha256Hex(stableStringify(baseCertificate)),
  };
}

/**
 * Handles the gnosis serialize Certificate workflow.
 */
export function serializeCertificate(cert: ClockworkCertificate): string {
  return JSON.stringify(cert, null, 2);
}

/**
 * Handles the gnosis serialize Universe Certificate workflow.
 */
export function serializeUniverseCertificate(
  certificate: UniverseCertificate
): string {
  return JSON.stringify(certificate, null, 2);
}

/**
 * Handles the gnosis deserialize Certificate workflow.
 */
export function deserializeCertificate(json: string): ClockworkCertificate {
  const parsed = JSON.parse(json) as ClockworkCertificate;
  if (parsed.version !== 1) {
    throw new Error(`Unknown certificate version: ${parsed.version}`);
  }
  if (
    parsed.sufficientStatisticTheorem !==
    'THM-VOID-BOUNDARY-SUFFICIENT-STATISTIC'
  ) {
    throw new Error(
      'Certificate missing sufficient statistic theorem reference'
    );
  }
  return parsed;
}

/**
 * Handles the gnosis deserialize Universe Certificate workflow.
 */
export function deserializeUniverseCertificate(
  json: string
): UniverseCertificate {
  const parsed = JSON.parse(json) as UniverseCertificate;
  if (parsed.version !== 1) {
    throw new Error(`Unknown universe certificate version: ${parsed.version}`);
  }
  const audited = auditUniverseCertificate(parsed);
  if (!audited.selfConsistent) {
    throw new Error(audited.issues.join('; '));
  }
  return parsed;
}

/**
 * Handles the gnosis audit Certificate workflow.
 */
export function auditCertificate(cert: ClockworkCertificate): {
  selfConsistent: boolean;
  issues: string[];
} {
  const issues: string[] = [];
  const sumCounts = cert.voidBoundary.counts.reduce(
    (sum, count) => sum + count,
    0
  );

  if (sumCounts !== cert.voidBoundary.totalEntries) {
    issues.push(
      `Void boundary count sum ${sumCounts} != totalEntries ${cert.voidBoundary.totalEntries}`
    );
  }

  const residual =
    cert.conservation.vFork - cert.conservation.wFold - cert.conservation.qVent;
  if (cert.conservation.conserved && Math.abs(residual) > 1e-10) {
    issues.push(`Conservation claimed but residual = ${residual}`);
  }

  if (cert.status === 'verified' && cert.deficit.openNodes > 0) {
    issues.push(
      `Status is verified but ${cert.deficit.openNodes} nodes have open deficit`
    );
  }

  if (cert.status === 'verified' && cert.checks.some((check) => !check.pass)) {
    issues.push('Status is verified but one or more checks failed');
  }

  if (cert.laplace) {
    const expectedWeights = cert.voidBoundary.counts.map((count) =>
      certificateBuleyeanWeight(cert.laplace!.rounds, count)
    );
    const finiteBoundary =
      isSafeNatural(cert.voidBoundary.counts.length) &&
      cert.voidBoundary.counts.every((count) => isSafeNatural(count));
    const computableWeights = expectedWeights.every(
      (weight) => Number.isSafeInteger(weight) && weight > 0
    );
    const withinZone =
      Number.isSafeInteger(cert.laplace.ceiling) &&
      cert.laplace.ceiling >= 1 &&
      expectedWeights.every(
        (weight) =>
          weight >= cert.laplace!.floor && weight <= cert.laplace!.ceiling
      );
    const finiteConjunction =
      Number.isSafeInteger(cert.checks.length) && cert.checks.length > 0;
    const finiteStages =
      isSafeNatural(cert.laplace.rounds) &&
      withinZone &&
      cert.conservation.conserved;

    if (cert.laplace.floor !== 1) {
      issues.push(
        `Laplace witness floor mismatch: expected 1, received ${cert.laplace.floor}`
      );
    }

    if (cert.laplace.ceiling !== cert.laplace.rounds + 1) {
      issues.push(
        `Laplace witness ceiling mismatch: expected ${
          cert.laplace.rounds + 1
        }, received ${cert.laplace.ceiling}`
      );
    }

    if (cert.laplace.gainCapacity !== cert.laplace.rounds) {
      issues.push(
        `Laplace witness gain mismatch: expected ${cert.laplace.rounds}, received ${cert.laplace.gainCapacity}`
      );
    }

    if (!sameNumberArray(cert.laplace.weights, expectedWeights)) {
      issues.push(
        `Laplace witness weights mismatch: expected [${expectedWeights.join(
          ', '
        )}], received [${cert.laplace.weights.join(', ')}]`
      );
    }

    if (cert.laplace.finiteBoundary !== finiteBoundary) {
      issues.push('Laplace witness finite-boundary flag mismatch');
    }

    if (cert.laplace.computableWeights !== computableWeights) {
      issues.push('Laplace witness computable-weights flag mismatch');
    }

    if (cert.laplace.withinZone !== withinZone) {
      issues.push('Laplace witness floor/ceiling zone flag mismatch');
    }

    if (cert.laplace.finiteConjunction !== finiteConjunction) {
      issues.push('Laplace witness finite-conjunction flag mismatch');
    }

    if (cert.laplace.finiteStages !== finiteStages) {
      issues.push('Laplace witness finite-stages flag mismatch');
    }
  }

  return { selfConsistent: issues.length === 0, issues };
}

function certificateBuleyeanWeight(
  rounds: number,
  rejectionCount: number
): number {
  return rounds - Math.min(rejectionCount, rounds) + 1;
}

function isSafeNatural(value: number): boolean {
  return Number.isSafeInteger(value) && value >= 0;
}

function sameNumberArray(
  left: readonly number[],
  right: readonly number[]
): boolean {
  return (
    left.length === right.length &&
    left.every((value, index) => value === right[index])
  );
}

/**
 * Handles the gnosis audit Universe Certificate workflow.
 */
export function auditUniverseCertificate(certificate: UniverseCertificate): {
  selfConsistent: boolean;
  issues: string[];
  law: ReturnType<typeof createLawState>;
} {
  const issues: string[] = [];
  const lowLevel = auditCertificate(certificate.clockwork);
  issues.push(...lowLevel.issues);

  const { digest, ...withoutDigest } = certificate;
  const expectedDigest = sha256Hex(stableStringify(withoutDigest));
  if (digest !== expectedDigest) {
    issues.push(
      `Universe certificate digest mismatch: expected ${expectedDigest}, received ${digest}`
    );
  }

  if (certificate.kernelId !== LAW_KERNEL_ID) {
    issues.push(
      `Universe certificate kernel mismatch: ${certificate.kernelId}`
    );
  }

  if (
    certificate.dimension !== 'dimensionless' ||
    certificate.currentDimension !== 'information'
  ) {
    issues.push('Universe certificate dimension envelope is invalid');
  }

  return {
    selfConsistent: issues.length === 0,
    issues,
    law: createLawState(certificate),
  };
}

// ── Composed Certificate (Topology + Code Quality) ──────────────

/**
 * Scanner certificate from mudras.us/api/void-certificate/:owner/:repo
 */
export interface ScannerCertificate {
  version: string;
  subject: string;
  scannedAt: string;
  scanCount: number;
  totalFindings: number;
  avgCrossingNumber: number;
  mergeRate: number;
  hash: string;
  issuer: string;
  theoremsBacking: string;
  verifyUrl: string;
}

/**
 * Unified trust proof: topological verification + code quality verification.
 * Two independent proofs composed into one certificate.
 */
export interface ComposedCertificate {
  topology: ClockworkCertificate;
  codeQuality: ScannerCertificate;
  composed: {
    subject: string;
    composedAt: number;
    topologyVerified: boolean;
    codeQualityClean: boolean;
    unified: boolean;
    totalTheoremsBacking: string;
  };
}

/**
 * Fetch a scanner certificate and compose it with a topological certificate.
 * Two independent verification domains, one unified trust proof.
 *
 * @param topoCert - ClockworkCertificate from gnosis topology verification
 * @param repoFullName - GitHub repo (owner/name) to fetch scanner cert for
 * @param scannerUrl - Base URL of the scanner
 */
export async function composeWithScannerCertificate(
  topoCert: ClockworkCertificate,
  repoFullName: string,
  scannerUrl = 'https://mudras.us'
): Promise<ComposedCertificate> {
  const resp = await fetch(`${scannerUrl}/api/void-certificate/${repoFullName}`);

  let scannerCert: ScannerCertificate;
  if (resp.ok) {
    const data = (await resp.json()) as { certificate: ScannerCertificate };
    scannerCert = data.certificate;
  } else {
    // Repo not yet scanned -- create a placeholder
    scannerCert = {
      version: '1.0',
      subject: repoFullName,
      scannedAt: '',
      scanCount: 0,
      totalFindings: -1,
      avgCrossingNumber: 0,
      mergeRate: 0,
      hash: '',
      issuer: 'mudras.us',
      theoremsBacking: 'not yet scanned',
      verifyUrl: `${scannerUrl}/repo/${repoFullName}`,
    };
  }

  return {
    topology: topoCert,
    codeQuality: scannerCert,
    composed: {
      subject: repoFullName,
      composedAt: Date.now(),
      topologyVerified: topoCert.status === 'verified',
      codeQualityClean: scannerCert.totalFindings === 0,
      unified: topoCert.status === 'verified' && scannerCert.totalFindings === 0,
      totalTheoremsBacking: '389 Lean 4 proofs (topology) + 147 scanner rules (code quality)',
    },
  };
}
