// @a0n/gnosis/src/universe/clockwork.ts
import { buleyeanWeight } from '../buleyean-kernel-shim.js';
import {
  fullBuleyeanAudit,
  type BuleyeanInput,
  type FullAuditReport,
} from './audit';
import {
  computeLawProgramMetrics,
  createLawState,
  LAW_DIMENSIONS,
  LAW_KERNEL_ID,
  UNIVERSE_CONSTANTS,
  UNIVERSE_FORMULAS,
  type LawKernel,
  type LawKernelEvaluation,
  type LawProgram,
  type LawEdge,
} from './kernel';
import {
  generateCertificate,
  generateUniverseCertificate,
} from './certificate';
import type {
  ClockworkVerdict,
  VerdictCheck,
  ConservationLedger,
  DeficitSummary,
  VoidSnapshot,
  HeatLedger,
  LaplaceWitness,
} from './verdict';
import {
  failCheck,
  passCheck,
  verdictStatus,
  verdictSummary,
} from './verdict';

interface VoidBoundary {
  counts: number[];
  totalEntries: number;
}

interface LocalDeficitNode {
  nodeId: string;
  localB1: number;
  isLeaf: boolean;
}

interface LocalDeficitReport {
  nodes: LocalDeficitNode[];
  totalDeficit: number;
  maxDeficit: number;
  totalVented: number;
}

interface DimensionEnvelope {
  readonly applicationDimension: string;
  readonly currentDimension: string;
}

export interface ClockworkConfig {
  readonly auditSevenLaws?: boolean;
  readonly computeHeat?: boolean;
  readonly verifyConservation?: boolean;
  readonly subject?: string;
}

const DEFAULT_CONFIG: Required<ClockworkConfig> = {
  auditSevenLaws: true,
  computeHeat: true,
  verifyConservation: true,
  subject: 'topology',
};

/**
 * Handles the gnosis evaluate Law Program workflow.
 */
export function evaluateLawProgram(
  program: LawProgram,
  config: ClockworkConfig = {}
): LawKernelEvaluation {
  const verdict = verify(program, config);
  const subject = config.subject ?? program.subject ?? DEFAULT_CONFIG.subject;
  const clockworkCertificate = generateCertificate(subject, verdict);
  const certificate = generateUniverseCertificate(
    program,
    verdict,
    clockworkCertificate
  );
  const law = createLawState(certificate);

  return {
    verdict,
    clockworkCertificate,
    certificate,
    law,
  };
}

export const GOD_FORMULA_KERNEL: LawKernel = {
  id: LAW_KERNEL_ID,
  verify(program: LawProgram): LawKernelEvaluation {
    return evaluateLawProgram(program);
  },
  constants: UNIVERSE_CONSTANTS,
  formulas: UNIVERSE_FORMULAS,
};

/**
 * Handles the gnosis verify workflow.
 */
export function verify(
  program: LawProgram,
  config: ClockworkConfig = {}
): ClockworkVerdict {
  const cfg = { ...DEFAULT_CONFIG, ...config };
  const checks: VerdictCheck[] = [];
  const dimensionEnvelope = readDimensionEnvelope(program);

  const metrics = computeLawProgramMetrics(program);
  const boundary = buildVoidBoundary(program, metrics.dimensionCount);
  const snapshot = readVoidSnapshot(boundary);

  if (snapshot.positivityHolds) {
    checks.push(
      passCheck(
        'positivity',
        'Buleyean positivity (sliver)',
        'THM-BULEYEAN-POSITIVITY'
      )
    );
  } else {
    checks.push(
      failCheck(
        'positivity',
        'Buleyean positivity (sliver)',
        'THM-BULEYEAN-POSITIVITY',
        'Some dimension has weight < 1'
      )
    );
  }

  const deficitReport = analyzeDeficit(program);
  const deficit = buildDeficitSummary(deficitReport);
  if (deficit.openNodes === 0) {
    checks.push(
      passCheck(
        'deficit-zero',
        'Zero deficit at all terminals',
        'THM-ZERO-DEFICIT-PRESERVES-INFORMATION'
      )
    );
  } else {
    checks.push(
      failCheck(
        'deficit-open',
        'Open deficit at terminals',
        'THM-DEFICIT-INFORMATION-LOSS',
        `${deficit.openNodes} nodes have β₁ > 0 (total deficit = ${deficit.totalDeficit})`
      )
    );
  }

  const conservation = cfg.verifyConservation
    ? checkConservation(program)
    : { vFork: 0, wFold: 0, qVent: 0, residual: 0, conserved: true };

  if (conservation.conserved) {
    checks.push(
      passCheck(
        'conservation',
        'First law conservation',
        'THM-FIRST-LAW-GENERAL'
      )
    );
  } else {
    checks.push(
      failCheck(
        'conservation',
        'First law conservation',
        'THM-FIRST-LAW-GENERAL',
        `V_fork(${conservation.vFork}) != W_fold(${conservation.wFold}) + Q_vent(${conservation.qVent}), residual = ${conservation.residual}`
      )
    );
  }

  const heat = cfg.computeHeat
    ? computeHeatLedger(program)
    : { totalHeat: 0, mergePoints: 0 };
  if (heat.totalHeat >= 0) {
    checks.push(
      passCheck('landauer', 'Landauer heat non-negative', 'THM-FOLD-HEAT')
    );
  } else {
    checks.push(
      failCheck(
        'landauer',
        'Landauer heat non-negative',
        'THM-FOLD-HEAT',
        `Negative heat ${heat.totalHeat}`
      )
    );
  }

  if (
    cfg.auditSevenLaws &&
    boundary.totalEntries > 0 &&
    shouldRunBuleyeanAudit(program)
  ) {
    const input: BuleyeanInput = {
      rounds: Math.max(boundary.totalEntries, 1),
      rejections: [...boundary.counts],
    };
    appendAuditChecks(fullBuleyeanAudit(input), checks);
  }

  if (!detectCycle(program)) {
    checks.push(passCheck('acyclicity', 'DAG acyclicity', 'THM-C1C4'));
  } else {
    checks.push(
      failCheck(
        'acyclicity',
        'DAG acyclicity',
        'THM-C1C4',
        'Topology contains a cycle'
      )
    );
  }

  const foldCount = countEdgeType(program, 'FOLD');
  const forkCount = countEdgeType(program, 'FORK');
  if (foldCount > 0) {
    checks.push(
      passCheck(
        'fold-erasure',
        'Non-injective fold erases information',
        'THM-FOLD-ERASURE'
      )
    );
  }
  if (forkCount > 0 && foldCount > 0) {
    checks.push(
      passCheck(
        'completeness',
        'Fork/fold DAG completeness',
        'THM-COMPLETENESS-DAG'
      )
    );
  }

  if (dimensionEnvelope.applicationDimension === 'dimensionless') {
    checks.push(
      passCheck(
        'dimensionless-app',
        'Applications stay dimensionless',
        'THM-ASCENSION-INFORMATION-LEVEL'
      )
    );
  } else {
    checks.push(
      failCheck(
        'dimensionless-app',
        'Applications stay dimensionless',
        'THM-ASCENSION-INFORMATION-LEVEL',
        `Application dimension must be dimensionless, received '${dimensionEnvelope.applicationDimension}'`
      )
    );
  }

  if (
    LAW_DIMENSIONS.includes(
      dimensionEnvelope.currentDimension as (typeof LAW_DIMENSIONS)[number]
    )
  ) {
    checks.push(
      passCheck(
        'current-dimension',
        'Current dimension is explicit',
        'THM-DIMENSION-CURRENT-EXPLICIT'
      )
    );
  } else {
    checks.push(
      failCheck(
        'current-dimension',
        'Current dimension is explicit',
        'THM-DIMENSION-CURRENT-EXPLICIT',
        `Current dimension '${dimensionEnvelope.currentDimension}' is not part of the universe lattice`
      )
    );
  }

  const laplaceCheckCount = 5;
  const laplace = buildLaplaceWitness(
    boundary,
    boundary.totalEntries,
    conservation.conserved,
    checks.length + laplaceCheckCount
  );
  appendLaplaceChecks(laplace, checks);

  const passed = checks.filter((check) => check.pass).length;
  const failed = checks.length - passed;
  const status = verdictStatus(checks);

  return {
    status,
    summary: verdictSummary(status, deficit, conservation),
    void: snapshot,
    deficit,
    conservation,
    heat,
    laplace,
    checks,
    passed,
    failed,
    verifiedAt: Date.now(),
  };
}

/**
 * Handles the gnosis verify Buleyean workflow.
 */
export function verifyBuleyean(
  rounds: number,
  rejections: number[]
): ClockworkVerdict {
  const boundary: VoidBoundary = {
    counts: [...rejections],
    totalEntries: rejections.reduce((sum, rejection) => sum + rejection, 0),
  };
  const snapshot = readVoidSnapshot(boundary, rounds);
  const checks: VerdictCheck[] = [];

  if (snapshot.positivityHolds) {
    checks.push(
      passCheck(
        'positivity',
        'Buleyean positivity (sliver)',
        'THM-BULEYEAN-POSITIVITY'
      )
    );
  } else {
    checks.push(
      failCheck(
        'positivity',
        'Buleyean positivity (sliver)',
        'THM-BULEYEAN-POSITIVITY',
        'Weight < 1 found'
      )
    );
  }

  const audit = fullBuleyeanAudit({
    rounds: Math.max(rounds, 1),
    rejections,
  });
  appendAuditChecks(audit, checks);

  const deficitValue = teleportDeficit(boundary);
  const deficit: DeficitSummary = {
    totalDeficit: deficitValue,
    maxDeficit: deficitValue,
    openNodes: deficitValue > 0 ? 1 : 0,
    totalNodes: 1,
    totalVented: 0,
    perNode: new Map([['boundary', deficitValue]]),
  };
  if (deficitValue === 0) {
    checks.push(
      passCheck(
        'deficit-zero',
        'Zero teleport deficit',
        'THM-ZERO-DEFICIT-PRESERVES-INFORMATION'
      )
    );
  } else {
    checks.push(
      failCheck(
        'deficit-open',
        'Open teleport deficit',
        'THM-DEFICIT-INFORMATION-LOSS',
        `Deficit = ${deficitValue}`
      )
    );
  }

  const conservation: ConservationLedger = {
    vFork: 0,
    wFold: 0,
    qVent: 0,
    residual: 0,
    conserved: true,
  };
  const heat: HeatLedger = {
    totalHeat: 0,
    mergePoints: 0,
  };
  const laplaceCheckCount = 5;
  const laplace = buildLaplaceWitness(
    boundary,
    rounds,
    conservation.conserved,
    checks.length + laplaceCheckCount
  );
  appendLaplaceChecks(laplace, checks);
  const passed = checks.filter((check) => check.pass).length;
  const failed = checks.length - passed;
  const status = verdictStatus(checks);

  return {
    status,
    summary: verdictSummary(status, deficit, conservation),
    void: snapshot,
    deficit,
    conservation,
    heat,
    laplace,
    checks,
    passed,
    failed,
    verifiedAt: Date.now(),
  };
}

function createVoidBoundary(dimensions: number): VoidBoundary {
  return {
    counts: new Array(Math.max(1, dimensions)).fill(0),
    totalEntries: 0,
  };
}

function shouldRunBuleyeanAudit(program: LawProgram): boolean {
  if (program.metadata?.buleyeanAudit === true) {
    return true;
  }

  if (program.edges.some((edge: LawEdge) => edge.type === 'REJECT')) {
    return true;
  }

  for (const node of program.nodes.values()) {
    if (node.properties.buleyeanAudit === 'true') {
      return true;
    }
  }

  for (const edge of program.edges) {
    if (edge.properties.buleyeanAudit === 'true') {
      return true;
    }
  }

  return false;
}

function readDimensionEnvelope(program: LawProgram): DimensionEnvelope {
  if (!program) {
    return {
      applicationDimension: 'dimensionless',
      currentDimension: 'information',
    };
  }
  const candidates: Array<Readonly<Record<string, string>>> = [
    ...(program.metadata
      ? [
          Object.fromEntries(
            Object.entries(program.metadata).flatMap(([key, value]) =>
              typeof value === 'string' ? [[key, value]] : []
            )
          ),
        ]
      : []),
    ...[...program.nodes.values()].map((node) => node.properties),
    ...program.edges.map((edge: LawEdge) => edge.properties),
  ];

  const applicationDimension =
    candidates
      .map(
        (candidate) =>
          candidate.applicationDimension ??
          candidate.appDimension ??
          candidate.dimensionless
      )
      .find((value) => typeof value === 'string' && value.length > 0) ??
    'dimensionless';

  const currentDimension =
    candidates
      .map(
        (candidate) =>
          candidate.currentDimension ??
          candidate.dimensionCurrent ??
          candidate.dimension ??
          candidate.dimensionMode
      )
      .find((value) => typeof value === 'string' && value.length > 0) ??
    'information';

  return {
    applicationDimension,
    currentDimension,
  };
}

function updateVoidBoundary(
  boundary: VoidBoundary,
  dimensionIdx: number,
  magnitude: number = 1
): void {
  boundary.counts[dimensionIdx] = (boundary.counts[dimensionIdx] ?? 0) + magnitude;
  boundary.totalEntries += magnitude;
}

function buleyeanPositivity(
  boundary: VoidBoundary,
  rounds: number = boundary.totalEntries
): boolean {
  return boundary.counts.every((count) => buleyeanWeight(rounds, count) >= 1);
}

function teleportDeficit(boundary: VoidBoundary): number {
  if (boundary.counts.length === 0) {
    return 0;
  }

  const minCount = Math.min(...boundary.counts);
  return boundary.totalEntries - minCount * boundary.counts.length;
}

function buildVoidBoundary(
  program: LawProgram,
  dimensions: number
): VoidBoundary {
  const boundary = createVoidBoundary(dimensions);
  let dimensionIdx = 0;

  for (const edge of program.edges) {
    if (
      edge.type === 'FORK' ||
      edge.type === 'EVOLVE' ||
      edge.type === 'SUPERPOSE' ||
      edge.type === 'ENTANGLE'
    ) {
      for (let index = 0; index < edge.targetIds.length; index++) {
        updateVoidBoundary(boundary, dimensionIdx % boundary.counts.length, 1);
        dimensionIdx++;
      }
    }

    if (edge.type === 'VENT') {
      updateVoidBoundary(boundary, dimensionIdx % boundary.counts.length, 1);
    }

    if (edge.type === 'RACE' || edge.type === 'SLIVER') {
      const losers = Math.max(0, edge.sourceIds.length - edge.targetIds.length);
      for (let index = 0; index < losers; index++) {
        updateVoidBoundary(
          boundary,
          (dimensionIdx + index) % boundary.counts.length,
          1
        );
      }
    }

    if (edge.type === 'REJECT') {
      for (let index = 0; index < edge.sourceIds.length; index++) {
        updateVoidBoundary(boundary, dimensionIdx % boundary.counts.length, 1);
        dimensionIdx++;
      }
    }
  }

  return boundary;
}

function readVoidSnapshot(
  boundary: VoidBoundary,
  rounds?: number
): VoidSnapshot {
  const effectiveRounds =
    rounds != null && rounds > 0 ? rounds : Math.max(boundary.totalEntries, 1);
  return {
    counts: [...boundary.counts],
    totalEntries: boundary.totalEntries,
    dimensions: boundary.counts.length,
    teleportDeficit: teleportDeficit(boundary),
    positivityHolds: buleyeanPositivity(boundary, effectiveRounds),
  };
}

function analyzeDeficit(program: LawProgram): LocalDeficitReport {
  const nodeB1 = new Map<string, number>();
  for (const nodeId of program.nodes.keys()) {
    nodeB1.set(nodeId, 0);
  }

  let totalVented = 0;

  for (const edge of program.edges) {
    const sourceB1 =
      edge.sourceIds.length > 0
        ? Math.max(
            ...edge.sourceIds.map(
              (sourceId: string) => nodeB1.get(sourceId.trim()) ?? 0
            )
          )
        : 0;

    let nextB1 = sourceB1;
    if (
      edge.type === 'FORK' ||
      edge.type === 'EVOLVE' ||
      edge.type === 'SUPERPOSE' ||
      edge.type === 'ENTANGLE'
    ) {
      nextB1 = sourceB1 + edge.targetIds.length - 1;
    } else if (
      edge.type === 'FOLD' ||
      edge.type === 'COLLAPSE' ||
      edge.type === 'OBSERVE'
    ) {
      nextB1 = Math.max(0, sourceB1 - (edge.sourceIds.length - 1));
    } else if (edge.type === 'RACE' || edge.type === 'SLIVER') {
      nextB1 = Math.max(
        0,
        sourceB1 - Math.max(0, edge.sourceIds.length - edge.targetIds.length)
      );
    } else if (edge.type === 'VENT') {
      nextB1 = Math.max(0, sourceB1 - 1);
    }

    for (const targetId of edge.targetIds) {
      const existing = nodeB1.get(targetId.trim()) ?? 0;
      nodeB1.set(targetId.trim(), Math.max(existing, nextB1));
      totalVented += Math.abs(nextB1 - existing);
    }
  }

  const nodes: LocalDeficitNode[] = [];
  let totalDeficit = 0;
  let maxDeficit = 0;
  for (const [nodeId, localB1] of nodeB1) {
    const node = program.nodes.get(nodeId);
    if (!node) continue;
    const isSink = node.labels.some(l => ['Sink', 'Terminal', 'Drain'].includes(l));
    const hasOutgoing = program.edges.some((edge: LawEdge) =>
      edge.sourceIds.some((sourceId: string) => sourceId.trim() === nodeId)
    );
    const isLeaf = !hasOutgoing && !isSink;
    nodes.push({ nodeId, localB1, isLeaf });

    if (isLeaf && localB1 > 0) {
      totalDeficit += localB1;
      maxDeficit = Math.max(maxDeficit, localB1);
    }
  }

  return { nodes, totalDeficit, maxDeficit, totalVented };
}

function buildDeficitSummary(report: LocalDeficitReport): DeficitSummary {
  const perNode = new Map<string, number>();
  let openNodes = 0;
  for (const node of report.nodes) {
    perNode.set(node.nodeId, node.localB1);
    if (node.isLeaf && node.localB1 > 0) {
      openNodes++;
    }
  }

  return {
    totalDeficit: report.totalDeficit,
    maxDeficit: report.maxDeficit,
    openNodes,
    totalNodes: report.nodes.length,
    totalVented: report.totalVented,
    perNode,
  };
}

function checkConservation(program: LawProgram): ConservationLedger {
  let generated = 0;
  let dissipated = 0;

  // 1. Explicit edge conservation
  for (const edge of program.edges) {
    const s = edge.sourceIds.length;
    const t = edge.targetIds.length;

    switch (edge.type) {
      case 'FORK':
      case 'SUPERPOSE':
      case 'ENTANGLE':
      case 'PROCESS':
      case 'EVOLVE':
        generated += Math.max(0, t - s);
        break;
      case 'FOLD':
      case 'COLLAPSE':
      case 'OBSERVE':
        dissipated += Math.max(0, s - t);
        break;
      case 'VENT':
      case 'REJECT':
        if (edge.properties.condition !== 'corridor-reject') {
          if (t === 0) {
            dissipated += s;
          } else {
            generated += Math.max(0, t - s);
            dissipated += Math.max(0, s - t);
          }
        }
        break;
      case 'RACE':
      case 'SLIVER':
        if (s > 1) {
          dissipated += (s - 1);
        }
        break;
      default:
        break;
    }
  }

  // 2. Boundary conservation (Sources/Sinks)
  const inDegrees = new Map<string, number>();
  const outDegrees = new Map<string, number>();
  for (const nodeId of program.nodes.keys()) {
    inDegrees.set(nodeId, 0);
    outDegrees.set(nodeId, 0);
  }
  for (const edge of program.edges) {
    for (const s of edge.sourceIds) {
      const sid = s.trim();
      outDegrees.set(sid, (outDegrees.get(sid) ?? 0) + 1);
    }
    for (const t of edge.targetIds) {
      const tid = t.trim();
      inDegrees.set(tid, (inDegrees.get(tid) ?? 0) + 1);
    }
  }

  const sources = [...program.nodes.keys()].filter(id => (inDegrees.get(id) ?? 0) === 0);
  const sinks = [...program.nodes.keys()].filter(id => (outDegrees.get(id) ?? 0) === 0);
  
  generated += sources.length;
  dissipated += sinks.length;

  const residual = generated - dissipated;

  return {
    vFork: generated,
    wFold: dissipated,
    qVent: 0,
    residual,
    conserved: Math.abs(residual) < 1e-10,
  };
}

function computeHeatLedger(program: LawProgram): HeatLedger {
  let totalHeat = 0;
  let mergePoints = 0;

  for (const edge of program.edges) {
    if (
      edge.type === 'FOLD' ||
      edge.type === 'COLLAPSE' ||
      edge.type === 'OBSERVE'
    ) {
      const sourceCount = edge.sourceIds.length;
      if (sourceCount > 1) {
        totalHeat += Math.log2(sourceCount);
        mergePoints++;
      }
    }
  }

  return { totalHeat, mergePoints };
}

function appendAuditChecks(
  audit: FullAuditReport,
  checks: VerdictCheck[]
): void {
  for (const check of audit.allChecks) {
    checks.push(
      check.pass
        ? passCheck(check.id, check.name, check.leanTheorem)
        : failCheck(check.id, check.name, check.leanTheorem, check.message)
    );
  }
}

function buildLaplaceWitness(
  boundary: VoidBoundary,
  rounds: number,
  conserved: boolean,
  checkCount: number
): LaplaceWitness {
  const weights = boundary.counts.map((count) => buleyeanWeight(rounds, count));
  const finiteBoundary =
    isSafeNatural(boundary.counts.length) &&
    boundary.counts.every((count) => isSafeNatural(count));
  const computableWeights = weights.every(
    (weight) => Number.isSafeInteger(weight) && weight > 0
  );
  const withinZone =
    Number.isSafeInteger(rounds + 1) &&
    rounds + 1 >= 1 &&
    weights.every((weight) => weight >= 1 && weight <= rounds + 1);
  const finiteConjunction = Number.isSafeInteger(checkCount) && checkCount > 0;
  const finiteStages = isSafeNatural(rounds) && withinZone && conserved;

  return {
    rounds,
    floor: 1,
    ceiling: rounds + 1,
    gainCapacity: rounds,
    weights,
    finiteBoundary,
    computableWeights,
    finiteConjunction,
    finiteStages,
    withinZone,
  };
}

function appendLaplaceChecks(
  laplace: LaplaceWitness,
  checks: VerdictCheck[]
): void {
  if (laplace.finiteBoundary) {
    checks.push(
      passCheck(
        'finite-boundary',
        'Void boundary stays finite',
        'THM-INFINITY-CEILING-FINITE'
      )
    );
  } else {
    checks.push(
      failCheck(
        'finite-boundary',
        'Void boundary stays finite',
        'THM-INFINITY-CEILING-FINITE',
        'Void boundary must remain a finite natural-number vector'
      )
    );
  }

  if (laplace.computableWeights) {
    checks.push(
      passCheck(
        'computational-ceiling',
        'Weights stay computable',
        'THM-COMPUTATIONAL-CEILING'
      )
    );
  } else {
    checks.push(
      failCheck(
        'computational-ceiling',
        'Weights stay computable',
        'THM-COMPUTATIONAL-CEILING',
        `Weights escaped the natural-number surface: [${laplace.weights.join(
          ', '
        )}]`
      )
    );
  }

  if (laplace.finiteConjunction) {
    checks.push(
      passCheck(
        'finite-conjunction',
        'Verification stays finitely self-referential',
        'THM-SELF-REFERENCE-CEILING'
      )
    );
  } else {
    checks.push(
      failCheck(
        'finite-conjunction',
        'Verification stays finitely self-referential',
        'THM-SELF-REFERENCE-CEILING',
        'Verification surface must remain a finite conjunction'
      )
    );
  }

  if (laplace.finiteStages) {
    checks.push(
      passCheck(
        'finite-stages',
        'Finite stages preserve sliver and conservation',
        'THM-INFINITY-CEILING-STAGES'
      )
    );
  } else {
    checks.push(
      failCheck(
        'finite-stages',
        'Finite stages preserve sliver and conservation',
        'THM-INFINITY-CEILING-STAGES',
        'Finite-stage witness requires safe finite rounds, bounded weights, and conservation'
      )
    );
  }

  if (laplace.withinZone) {
    checks.push(
      passCheck(
        'floor-ceiling-sandwich',
        'Weights stay inside the floor/ceiling sandwich',
        'THM-FLOOR-CEILING-SANDWICH'
      )
    );
  } else {
    checks.push(
      failCheck(
        'floor-ceiling-sandwich',
        'Weights stay inside the floor/ceiling sandwich',
        'THM-FLOOR-CEILING-SANDWICH',
        `Weights escaped [${laplace.floor}, ${
          laplace.ceiling
        }]: [${laplace.weights.join(', ')}]`
      )
    );
  }
}

function isSafeNatural(value: number): boolean {
  return Number.isSafeInteger(value) && value >= 0;
}

function detectCycle(program: LawProgram): boolean {
  const adjacency = new Map<string, string[]>();
  const inDegree = new Map<string, number>();

  for (const nodeId of program.nodes.keys()) {
    adjacency.set(nodeId, []);
    inDegree.set(nodeId, 0);
  }

  for (const edge of program.edges) {
    for (const sourceId of edge.sourceIds) {
      const normalizedSource = sourceId.trim();
      if (!adjacency.has(normalizedSource)) {
        adjacency.set(normalizedSource, []);
      }
      if (!inDegree.has(normalizedSource)) {
        inDegree.set(normalizedSource, 0);
      }
      for (const targetId of edge.targetIds) {
        const normalizedTarget = targetId.trim();
        if (!inDegree.has(normalizedTarget)) {
          inDegree.set(normalizedTarget, 0);
        }
        adjacency.get(normalizedSource)?.push(normalizedTarget);
        inDegree.set(
          normalizedTarget,
          inDegree.get(normalizedTarget)! + 1
        );
      }
    }
  }

  const queue: string[] = [];
  for (const [nodeId, degree] of inDegree) {
    if (degree === 0) {
      queue.push(nodeId);
    }
  }

  let visited = 0;
  while (queue.length > 0) {
    const nodeId = queue.shift()!;
    visited++;
    for (const neighbor of adjacency.get(nodeId) ?? []) {
      const nextDegree = (inDegree.get(neighbor) ?? 1) - 1;
      inDegree.set(neighbor, nextDegree);
      if (nextDegree === 0) {
        queue.push(neighbor);
      }
    }
  }

  const totalNodes = new Set([...adjacency.keys(), ...inDegree.keys()]).size;
  return visited < totalNodes;
}

function countEdgeType(program: LawProgram, type: string): number {
  return program.edges.filter((edge: LawEdge) => edge.type === type).length;
}
