// @a0n/.ignored_gnosis/gnosis-time.ts
/**
 * Gnosis Time Core
 * Implementation of Gnostic Periodicity and the 12-Minute Aeon Clock
 *
 * This module provides the temporal primitives for human-manifold synchronization.
 * It implements the 'Aeon (12)' constant as the optimal periodic step for
 * cognitive focus and physiological recovery.
 */

export const GNOSIS_CONSTANTS = {
  AEON: 12, // The 12-minute sweet spot (micro-flow threshold)
  PRIMITIVE: 5, // The 5-minute primitive window
  HOUR: 60, // The standard Gnostic Hour (5 * 12)
  DAY: 1440, // The Solar Day (120 * 12)
};

/**
 * Validates if a duration (in minutes) is Topologically Stable.
 * A duration is stable if it divides the Gnostic Hour (60) or the Day (1440).
 */
export function isTopologicallyStable(minutes: number): boolean {
  if (minutes <= 0) return false;
  return GNOSIS_CONSTANTS.HOUR % minutes === 0 || GNOSIS_CONSTANTS.DAY % minutes === 0;
}

/**
 * The 12-12 Aeon Syzygy
 * A balanced work-rest cycle for high-performance manifold stability.
 */
export interface AeonSyzygy {
  workMinutes: number;
  restMinutes: number;
  multiplier: number;
}

export const AEON_SYZYGY_BASELINE: AeonSyzygy = {
  workMinutes: GNOSIS_CONSTANTS.AEON,
  restMinutes: GNOSIS_CONSTANTS.AEON,
  multiplier: 1.0,
};

/**
 * Scales a Gnostic duration based on the Developmental Overtone (Multiplier).
 * Used for child/elder synchronization where the Aeon 'stretches' or 'contracts'.
 */
export function scaleAeon(minutes: number, multiplier: number): number {
  return Math.round(minutes * multiplier);
}

/**
 * Returns the 'Cognitive State' for a given elapsed time.
 * Maps to the Attentional Blink and Micro-flow thresholds.
 */
export type CognitiveState = 'BUFFERING' | 'MICRO_FLOW' | 'BLINK_WARNING' | 'VIGILANCE_DECAY';

/**
 * Handles the gnosis get Cognitive State workflow.
 */
export function getCognitiveState(elapsedMinutes: number): CognitiveState {
  if (elapsedMinutes < 2) return 'BUFFERING';
  if (elapsedMinutes <= GNOSIS_CONSTANTS.AEON) return 'MICRO_FLOW';
  if (elapsedMinutes <= 15) return 'BLINK_WARNING';
  return 'VIGILANCE_DECAY';
}

/**
 * Gnostic Timer Preset
 * Pre-defined stable intervals for productivity and meditation.
 */
export interface GnosticTimerPreset {
  id: string;
  name: string;
  duration: number;
  type: 'FOCUS' | 'RECOVERY' | 'SYZYGY';
  description: string;
}

export const GNOSTIC_TIMER_PRESETS: GnosticTimerPreset[] = [
  {
    id: 'aeon-sprint',
    name: 'Aeon Focus Sprint',
    duration: 12,
    type: 'FOCUS',
    description: 'The optimal micro-flow threshold for complex sub-task resolution.',
  },
  {
    id: 'aeon-recovery',
    name: 'Aeon Recovery Window',
    duration: 12,
    type: 'RECOVERY',
    description: 'The Goldilocks zone for cortisol regulation and parasympathetic engagement.',
  },
  {
    id: 'syzygy-12-12',
    name: 'Aeon-Aeon Syzygy',
    duration: 24, // 12 + 12
    type: 'SYZYGY',
    description: 'A balanced work-rest cycle to prevent compounding exhaustion.',
  },
  {
    id: 'primitive-shield',
    name: 'Primitive Shield',
    duration: 5,
    type: 'FOCUS',
    description: 'A sharp 5-minute burst for high-arousal tasks.',
  },
];
