// @a0n/gnosis/distributed-inference-host/src/acoustic-source-counting.ts
/**
 * # Acoustic Source Counting — how many speakers are in the room
 *
 * Mirrors `Gnosis.Mesh.AcousticSourceCounting` from `open-source/gnosis-math`,
 * the Rustic Church carrier for *source counting*: from a window of 1-D acoustic
 * observations (each a scalar feature — a quantized pitch, bearing, or formant
 * centroid), estimate HOW MANY distinct sources (speakers) are present. The
 * estimate is not a free-floating number; it is the *length of an admitted
 * clustering*. The Skeptic admits a proposed source count only when the
 * observations partition into that many groups that are
 *
 *  - **internally tight** — every member of a group sits within `intraTol` of
 *    the group's representative (its head), so the group is one source's spread
 *    and not two sources smeared together;
 *  - **mutually separated** — any two distinct groups' representatives differ by
 *    more than `sepThreshold`, so the split is not a phantom (noise around one
 *    speaker carved into two); and
 *  - **supported** — every group has at least `minMembers` members, so it is a
 *    real source rather than a single stray observation.
 *
 * ## What is (and is not) carried
 *
 * Two failure modes are refuted, both about the *count*:
 *
 *  1. **Over-merge (under-count):** calling two well-separated speakers one
 *     source. The merged group's internal spread blows past `intraTol`, so the
 *     1-cluster proposal fails the tightness gate.
 *  2. **Over-split (over-count):** carving one speaker's jitter into several
 *     "sources". Two representatives sit within `sepThreshold`, so the proposal
 *     fails the separation gate — even when every fragment has support.
 *
 * This module is the runtime promotion check for a source-count observation. It
 * carries ONLY the finite contract the Lean file proves: the saturating feature
 * metric (`adist`), the three gates (`tightness` / `wellSeparated` / `supported`)
 * and their aggregate admission (`admitSourceCount`). Real diarization,
 * beamforming, embedding extraction, model-order selection, and overlapped-speech
 * handling stay runtime/empirical obligations OUTSIDE this certificate.
 *
 * The promotion check folds the contract into the same epistemic-outcome
 * vocabulary the sibling mesh carriers already speak
 * (`claim_admitted` / `observation_only` / `claim_withheld`), so a source-count
 * claim cannot be displayed until it has cleared the Lean-side gates.
 *
 * Representative convention: a group's representative is its head. The empty
 * group has no representative (`0`) and is treated as unsupported.
 *
 * ## Lean source-of-truth
 *
 * @see open-source/gnosis-math/Gnosis/Mesh/AcousticSourceCounting.lean
 * @see open-source/gnosis-math/Gnosis/Mesh/MicrophoneFingerprint.lean (mesh sibling)
 */

/* ------------------------------------------------------------------ */
/*  Feature distance, observations, clusters                          */
/* ------------------------------------------------------------------ */

/** An observation is a 1-D acoustic feature (a quantized pitch/bearing scalar). */
export type Observation = number;

/** A window of observations over a frame. */
export type Window = readonly Observation[];

/** A cluster (proposed source group) is a list of observations; its
 *  representative is its head. */
export type Cluster = readonly Observation[];

/** A proposed clustering: the list of groups. The admitted source count is its
 *  length. */
export type Clustering = readonly Cluster[];

/** Absolute difference on `Nat` via saturating subtraction (`|a - b|`). The 1-D
 *  acoustic feature residual metric, shared with the mesh corridor carriers.
 *  Mirrors `AcousticSourceCounting.adist`. */
export function adist(a: number, b: number): number {
  return Math.max(0, a - b) + Math.max(0, b - a);
}

/** The representative of a cluster: its head, or `0` for the empty cluster
 *  (which is independently rejected by the support gate). Mirrors `repOf`. */
export function repOf(group: Cluster): number {
  return group.length === 0 ? 0 : group[0];
}

/* ------------------------------------------------------------------ */
/*  Gate 1: intra-cluster tightness                                   */
/* ------------------------------------------------------------------ */

/**
 * A cluster is tight when every member is within `intraTol` of its
 * representative (its head). The merged-cluster over-count failure is exactly a
 * tightness failure. Mirrors `clusterTightB`.
 */
export function tightness(intraTol: number, group: Cluster): boolean {
  const rep = repOf(group);
  for (const a of group) {
    if (adist(a, rep) > intraTol) {
      return false;
    }
  }
  return true;
}

/** Every cluster in a clustering is tight. Mirrors `allTightB`. */
export function allTight(intraTol: number, clusters: Clustering): boolean {
  return clusters.every((g) => tightness(intraTol, g));
}

/* ------------------------------------------------------------------ */
/*  Gate 2: inter-cluster separation                                  */
/* ------------------------------------------------------------------ */

/** The list of representatives of a clustering. Mirrors `repsOf`. */
export function repsOf(clusters: Clustering): number[] {
  return clusters.map((g) => repOf(g));
}

/**
 * The clustering is well-separated when its representatives are pairwise more
 * than `sepThreshold` apart. Two clusters within `sepThreshold` mark an
 * over-split and fail this gate. Mirrors `wellSeparatedB`.
 */
export function wellSeparated(
  sepThreshold: number,
  clusters: Clustering,
): boolean {
  const reps = repsOf(clusters);
  for (let i = 0; i < reps.length; i++) {
    for (let j = i + 1; j < reps.length; j++) {
      if (!(adist(reps[i], reps[j]) > sepThreshold)) {
        return false;
      }
    }
  }
  return true;
}

/* ------------------------------------------------------------------ */
/*  Gate 3: support                                                   */
/* ------------------------------------------------------------------ */

/** A cluster has support when it has at least `minMembers` observations.
 *  Mirrors `supportedB`. */
export function supportedCluster(minMembers: number, group: Cluster): boolean {
  return group.length >= minMembers;
}

/** Every cluster in a clustering has support. Mirrors `allSupportedB`. */
export function supported(minMembers: number, clusters: Clustering): boolean {
  return clusters.every((g) => supportedCluster(minMembers, g));
}

/* ------------------------------------------------------------------ */
/*  Runtime promotion check (epistemic outcome)                       */
/* ------------------------------------------------------------------ */

/**
 * Epistemic outcome of a source-count observation. Identical vocabulary to the
 * sibling mesh carriers, kept as a local literal union so this certificate has
 * no import coupling:
 *
 *  - `claim_admitted`    — every cluster is tight, the clusters are mutually
 *                          well-separated, and every cluster has support, so the
 *                          source count (the clustering length) may be promoted.
 *  - `observation_only`  — reserved for the empty clustering: there is no count
 *                          to refute and none to certify, so it is reported,
 *                          never promoted.
 *  - `claim_withheld`    — a gate failed (an over-merge that is not tight, or an
 *                          over-split whose reps are too close, or an
 *                          unsupported fragment), so the proposed count is
 *                          refused.
 */
export type SourceCountEpistemicOutcome =
  | 'claim_admitted'
  | 'observation_only'
  | 'claim_withheld';

/** The reason a source-count claim was held back from promotion. */
export type SourceCountLimitationClass =
  | 'over_merge_undercount'
  | 'over_split_overcount'
  | 'unsupported_fragment'
  | 'empty_clustering'
  | 'none';

export interface SourceCountAdmission {
  readonly outcome: SourceCountEpistemicOutcome;
  readonly limitation: SourceCountLimitationClass;
  /** The proposed source count (the clustering length, displayable). */
  readonly count: number;
  /** Whether every cluster passed the tightness gate. */
  readonly tight: boolean;
  /** Whether the clusters passed the separation gate. */
  readonly separated: boolean;
  /** Whether every cluster passed the support gate. */
  readonly supported: boolean;
  readonly summary: string;
}

export interface SourceCountAdmitOptions {
  /** Max member spread to a cluster's representative (tightness gate). */
  readonly intraTol: number;
  /** Reps must differ by more than this (separation gate). */
  readonly sepThreshold: number;
  /** Minimum members for a cluster to count as a source (support gate). */
  readonly minMembers: number;
}

/**
 * The promotion check Monster Mesh runs before any UI surface displays a
 * source-count claim. It folds the Lean contract into an epistemic outcome:
 *
 *  - the tightness gate refutes an over-merge: a cluster smearing two speakers
 *    together is not tight, so the under-count is withheld (mirroring
 *    `over_merge_rejected`);
 *  - the separation gate refutes an over-split: representatives within
 *    `sepThreshold` carve one speaker into phantoms, so the over-count is
 *    withheld (mirroring `over_split_rejected`);
 *  - the support gate refutes unsupported fragments;
 *  - only a tight, separated, supported clustering is admitted, with the count
 *    set to the clustering length (mirroring `two_sources_admitted` /
 *    `two_sources_count`).
 *
 * The empty clustering is the degenerate observation-only case: no count to
 * certify, no count to refute.
 *
 * @param clusters the proposed clustering (the source partition under examination)
 * @param opts      the three gate thresholds
 */
export function admitSourceCount(
  clusters: Clustering,
  opts: SourceCountAdmitOptions,
): SourceCountAdmission {
  const count = clusters.length;
  const isTight = allTight(opts.intraTol, clusters);
  const isSeparated = wellSeparated(opts.sepThreshold, clusters);
  const isSupported = supported(opts.minMembers, clusters);

  if (count === 0) {
    return {
      outcome: 'observation_only',
      limitation: 'empty_clustering',
      count,
      tight: isTight,
      separated: isSeparated,
      supported: isSupported,
      summary: 'empty clustering: no source count to certify or refute',
    };
  }

  // Tightness gate: an over-merge crams well-separated speakers into one cluster
  // whose internal spread exceeds intraTol — refuse the under-count.
  if (!isTight) {
    return {
      outcome: 'claim_withheld',
      limitation: 'over_merge_undercount',
      count,
      tight: isTight,
      separated: isSeparated,
      supported: isSupported,
      summary: `over-merge: a cluster's internal spread exceeds intraTol ${opts.intraTol} — proposed count ${count} under-counts the sources`,
    };
  }

  // Separation gate: an over-split carves one speaker's jitter into fragments
  // whose reps sit within sepThreshold — refuse the over-count.
  if (!isSeparated) {
    return {
      outcome: 'claim_withheld',
      limitation: 'over_split_overcount',
      count,
      tight: isTight,
      separated: isSeparated,
      supported: isSupported,
      summary: `over-split: two cluster representatives are within sepThreshold ${opts.sepThreshold} — proposed count ${count} over-counts the sources`,
    };
  }

  // Support gate: every cluster must have enough members to be a real source.
  if (!isSupported) {
    return {
      outcome: 'claim_withheld',
      limitation: 'unsupported_fragment',
      count,
      tight: isTight,
      separated: isSeparated,
      supported: isSupported,
      summary: `unsupported: a cluster has fewer than minMembers ${opts.minMembers} observations — not a real source`,
    };
  }

  return {
    outcome: 'claim_admitted',
    limitation: 'none',
    count,
    tight: isTight,
    separated: isSeparated,
    supported: isSupported,
    summary: `tight, well-separated, supported clustering — source count ${count} certified`,
  };
}

/* ------------------------------------------------------------------ */
/*  Concrete fixtures (mirror the Lean fixtures)                      */
/* ------------------------------------------------------------------ */

/** The acoustic window: three observations near 10-12 (one speaker) and three
 *  near 40-42 (a second speaker). Mirrors `AcousticSourceCounting.window`. */
export const window: Window = [10, 11, 12, 40, 41, 42];

/** Intra-cluster tolerance (max member spread to representative). */
export const intraTol = 5;

/** Inter-cluster separation threshold (reps must differ by more than this). */
export const sepThreshold = 10;

/** Minimum members for a cluster to count as a source. */
export const minMembers = 1;

/** The correct 2-source grouping: two tight, well-separated, supported clusters.
 *  Admitted count 2. Mirrors `twoClusters`. */
export const twoClusters: Clustering = [
  [10, 11, 12],
  [40, 41, 42],
];

/** The over-merged 1-source grouping of the SAME window: both speakers crammed
 *  into one cluster whose internal spread (32) blows past `intraTol`. Rejected.
 *  Mirrors `mergedOne`. */
export const mergedOne: Clustering = [[10, 11, 12, 40, 41, 42]];

/** The over-split grouping of the SAME window: the first speaker's jitter carved
 *  into two fragments whose representatives (10 and 12) sit within `sepThreshold`.
 *  Rejected on separation, not support. Mirrors `overSplit`. */
export const overSplit: Clustering = [[10, 11], [12], [40, 41, 42]];

/**
 * Acoustic Source Counting master witness. Bundles the boundary facts the Lean
 * theorem `source_counting_master` proves, each recomputed here from first
 * principles (never imported from Lean IR).
 *
 * Mirrors `AcousticSourceCounting.source_counting_master`.
 */
export const AcousticSourceCountingMaster = {
  /** The two honest clusters are each internally tight (spreads 2 and 2 <= 5). */
  twoClustersTight: allTight(intraTol, twoClusters) === true,

  /** The merged cluster is NOT tight: its spread (32) exceeds `intraTol`. */
  mergedOneNotTight: allTight(intraTol, mergedOne) === false,

  /** The two honest clusters are well-separated (reps 10 vs 40, gap 30 > 10). */
  twoClustersSeparated: wellSeparated(sepThreshold, twoClusters) === true,

  /** The over-split clustering is NOT well-separated (reps 10 and 12, gap 2 <= 10). */
  overSplitNotSeparated: wellSeparated(sepThreshold, overSplit) === false,

  /** Every honest cluster has support (lengths 3,3 >= 1). */
  twoClustersSupported: supported(minMembers, twoClusters) === true,

  /** Every over-split fragment has support too: the over-split is refuted by
   *  separation, NOT by lack of support. */
  overSplitSupported: supported(minMembers, overSplit) === true,

  /** The correct 2-source grouping is admitted with count 2. */
  twoSourcesAdmitted:
    admitSourceCount(twoClusters, { intraTol, sepThreshold, minMembers })
      .outcome === 'claim_admitted',

  /** The admitted count for the honest grouping is 2. */
  twoSourcesCount:
    admitSourceCount(twoClusters, { intraTol, sepThreshold, minMembers })
      .count === 2,

  /** Over-merge rejection: the 1-cluster proposal fails the tightness gate
   *  (under-count). */
  overMergeRejected:
    admitSourceCount(mergedOne, { intraTol, sepThreshold, minMembers })
      .outcome === 'claim_withheld',

  /** Over-split rejection: the 3-cluster proposal fails the separation gate
   *  (over-count), and on separation, not support. */
  overSplitRejected:
    admitSourceCount(overSplit, { intraTol, sepThreshold, minMembers })
      .outcome === 'claim_withheld',
} as const;
