// @a0n/gnosis/distributed-inference-host/src/agentic-chat.ts
/**
 * Runtime-neutral agentic chat loop for distributed inference hosts.
 *
 * The local model stays responsible for text generation. This module adds the
 * orchestration around it: superserver MCP preflight, tool prompt injection,
 * tool execution, light C0-C3 metacognitive gates, and OpenAI-compatible
 * response shaping.
 */

import {
  renderThothGateTelemetryPromptContext,
  thothGateTelemetryActionHint,
  thothGateTelemetryTotalPressure,
  type ThothGateTelemetryMetrics,
} from './thoth-gate-telemetry';
import {
  renderThothNpcHalogramPromptContext,
  type ThothNpcHalogramMetrics,
} from './thoth-npc-halogram';
import {
  renderThothNearbyBusinessesPromptContext,
  renderThothPlacePointerPromptContext,
  type ThothNearbyBusinessesMetrics,
} from './thoth-nearby-businesses';
import {
  createThothOathSelfManagementMcpClient,
  type ThothOathLedgerStore,
} from './thoth-oath-tools';
import { createThothPlaceMcpClient } from './thoth-place-tools';
import type { ThothPlaceSimHost } from './thoth-place-sim';
import {
  AgentEconomy,
  classifyToolRisk,
  economyStatusBlock,
  type EconomySummary,
} from './agentic-economy';
import {
  formatNativeToolsDeclaration,
  parseNativeToolCalls,
  type NativeToolFamily,
} from './native-tool-formats';

export type AgenticChatRole =
  | 'system'
  | 'developer'
  | 'user'
  | 'assistant'
  | 'tool';

export interface AgenticToolCall {
  id: string;
  type: 'function';
  function: {
    name: string;
    arguments: string;
  };
}

export interface AgenticChatMessage {
  role: AgenticChatRole;
  content: string | null;
  tool_calls?: AgenticToolCall[];
  tool_call_id?: string;
  name?: string;
}

export interface ToolActivationMetadata {
  patterns: string[];
  keywords?: string[];
  onlyRules?: boolean;
  agentResponse?: boolean;
  tier?: number;
  skipReview?: boolean;
  threshold?: number;
  argExtractors?: Record<string, string>;
  prerequisites?: Array<{
    toolName: string;
    argMapping: Record<string, string>;
  }>;
}

export interface AgenticToolDefinition {
  type: 'function';
  function: {
    name: string;
    description?: string;
    parameters?: {
      type: 'object';
      properties: Record<string, unknown>;
      required?: string[];
    };
  };
  activation?: ToolActivationMetadata;
  source?: 'superserver' | 'provided';
  metadata?: Record<string, unknown>;
}

export type AgenticToolChoice =
  | 'auto'
  | 'none'
  | 'required'
  | { type: 'function'; function: { name: string } };

export interface AgenticChatRequest {
  model: string;
  messages: AgenticChatMessage[];
  temperature?: number;
  max_tokens?: number;
  tools?: AgenticToolDefinition[];
  tool_choice?: AgenticToolChoice;
  auto_tools?: boolean;
  execute_tools?: boolean;
  max_tool_rounds?: number;
  evaluate_intent?: boolean;
  intent_confidence_threshold?: number;
  max_continuation_nudges?: number;
  response_format?: { type: 'text' | 'json_object' } | null;
  thothGateTelemetry?: ThothGateTelemetryMetrics;
  thothNpcHalogram?: ThothNpcHalogramMetrics;
  thothNearbyBusinesses?: ThothNearbyBusinessesMetrics;
  /** Enable the thoth/place.* sense-of-place tools (needs
   *  runtime.thothPlaceSimHost). */
  thothPlaceTools?: boolean;
  /** Request observer that seeds the default city for the place tools. */
  thothPlaceObserver?: {
    lat: number;
    lon: number;
    cityName?: string;
  } | null;
  metacognition?: {
    enabled?: boolean;
    reviewToolActivations?: boolean;
    validateToolResults?: boolean;
    maxValidationRetries?: number;
  };
  /** Cap on how many tool descriptions to embed in the system prompt. When
   *  visibleTools exceeds this, only the MCP discovery tools (network/search,
   *  network/directory, network/docs) are injected so the model can find
   *  what it needs on demand without blowing its context window. */
  max_tools_in_prompt?: number;
}

export interface TextGenerationRequest {
  model: string;
  messages: AgenticChatMessage[];
  temperature?: number;
  maxTokens?: number;
  responseFormat?: { type: 'text' | 'json_object' } | null;
  round: number;
  purpose:
    | 'chat'
    | 'tool-review'
    | 'tool-args'
    | 'intent-evaluation'
    | 'tool-retry';
  /** Present when tools are visible this round. Backends without a native
   *  tool-calling format (the default) ignore this; the Mistral V3 template
   *  needs it directly (its `[AVAILABLE_TOOLS]` segment sits before the first
   *  `[INST]` turn, outside any message's content, so it can't be spliced in
   *  via `messages` the way Qwen's/Gemma4's declaration text is). */
  tools?: AgenticToolDefinition[];
}

export interface TextGenerationResult {
  content: string;
  promptTokens?: number;
  completionTokens?: number;
  model?: string;
}

export interface ToolExecutionResult {
  toolCallId: string;
  toolName: string;
  args: Record<string, unknown>;
  result: unknown;
  success: boolean;
  durationMs: number;
  error?: string;
}

export interface AgenticMcpClient {
  readonly sessionId?: string | null;
  preflight?(): Promise<McpPreflightResult>;
  listTools(options?: { forceRefresh?: boolean }): Promise<AgenticToolDefinition[]>;
  callTool(
    name: string,
    args: Record<string, unknown>,
    options?: { toolCallId?: string },
  ): Promise<ToolExecutionResult>;
}

export interface AgenticChatRuntime {
  infer(request: TextGenerationRequest): Promise<TextGenerationResult>;
  mcp?: AgenticMcpClient;
  thothOathLedgerStore?: ThothOathLedgerStore;
  /** Headless living-city sim host backing the thoth/place.* tools. */
  thothPlaceSimHost?: ThothPlaceSimHost;
  logger?: Pick<Console, 'debug' | 'info' | 'warn' | 'error'>;
  now?: () => number;
  randomId?: (prefix: string) => string;
  /** Native per-family tool-calling wire format this runtime's chat template
   *  supports (resolved once from `chatTemplate`, see `native-tool-formats.ts`),
   *  or `null`/unset to keep today's generic fenced-block convention. */
  nativeToolFamily?: NativeToolFamily | null;
}

export interface McpPreflightResult {
  tools: AgenticToolDefinition[];
  sessionId?: string | null;
  durationMs: number;
  cached: boolean;
}

export interface SuperserverMcpClientConfig {
  mcpUrl?: string;
  fetcher?: Pick<typeof globalThis, 'fetch'>['fetch'] | { fetch: typeof fetch };
  authToken?: string;
  sessionId?: string;
  timeoutMs?: number;
  cacheTtlMs?: number;
}

export interface AgenticTraceEntry {
  layer: 'preflight' | 'C0' | 'C1' | 'C2' | 'C3' | 'C4' | 'intent' | 'loop';
  message: string;
  data?: Record<string, unknown>;
}

export interface AgenticChatCompletionResponse {
  id: string;
  object: 'chat.completion';
  created: number;
  model: string;
  choices: Array<{
    index: number;
    message: {
      role: 'assistant';
      content: string | null;
      tool_calls?: AgenticToolCall[];
    };
    finish_reason: 'stop' | 'tool_calls' | 'length';
  }>;
  usage: {
    prompt_tokens: number;
    completion_tokens: number;
    total_tokens: number;
  };
  tools_used?: string[];
  agentic: {
    rounds: number;
    toolPreflight: {
      enabled: boolean;
      toolCount: number;
      visibleToolCount: number;
      cached: boolean;
      durationMs: number;
      sessionId?: string | null;
    };
    metacognitionTrace: AgenticTraceEntry[];
    nudges: number;
    maxToolRoundsReached: boolean;
    /** Aeon-12 tool economy state at the end of this request — see `agentic-economy.ts`. */
    economy: EconomySummary;
  };
}

interface RpcEnvelope {
  jsonrpc: '2.0';
  id: string | number | null;
  method: string;
  params?: Record<string, unknown>;
}

interface RpcResponse {
  jsonrpc?: '2.0';
  id?: string | number | null;
  result?: unknown;
  error?: { code?: number; message?: string; data?: unknown };
}

interface McpContentBlock {
  type: string;
  text?: string;
}

interface McpToolWire {
  name: string;
  description?: string;
  inputSchema?: {
    type?: string;
    properties?: Record<string, unknown>;
    required?: string[];
  };
  activation?: ToolActivationMetadata;
  [key: string]: unknown;
}

interface ToolCandidate {
  toolName: string;
  score: number;
  matchedPatterns: string[];
  suggestedArgs: Record<string, unknown>;
  onlyRules?: boolean;
  skipReview?: boolean;
  agentResponse?: boolean;
  tier?: number;
}

interface ActivationResult {
  candidates: ToolCandidate[];
  topCandidate: ToolCandidate | null;
  entropy: number;
  autoActivate: boolean;
}

interface IntentEvaluation {
  satisfied: boolean;
  confidence: number;
  reasoning: string;
  suggestedNextAction?: string;
}

interface ResultValidation {
  valid: boolean;
  quality: number;
  issues: string[];
  suggestRetry: boolean;
  modifiedArgs?: Record<string, unknown>;
}

const MCP_SESSION_HEADER = 'mcp-session-id';
const DEFAULT_MCP_URL = 'https://mcp.forkjoin.ai/mcp';
const DEFAULT_MCP_TIMEOUT_MS = 20_000;
const DEFAULT_TOOL_CACHE_TTL_MS = 5 * 60 * 1000;
const DEFAULT_MAX_TOOL_ROUNDS = 8;
const DEFAULT_MAX_VISIBLE_TOOLS = 32;
const DEFAULT_INTENT_CONFIDENCE = 0.7;
const DEFAULT_MAX_NUDGES = 2;
const DEFAULT_AUTO_ACTIVATE_THRESHOLD = 0.75;
const AUTO_ACTIVATE_MAX_ENTROPY = 0.4;

const PINNED_GATEWAY_TOOLS = new Set([
  'network/search',
  'network/directory',
  'network/docs',
  'network/matrix',
]);

const STOP_WORDS = new Set([
  'a',
  'an',
  'and',
  'are',
  'as',
  'at',
  'be',
  'by',
  'for',
  'from',
  'in',
  'into',
  'is',
  'of',
  'on',
  'or',
  'the',
  'to',
  'with',
]);

export class SuperserverMcpClient implements AgenticMcpClient {
  private readonly mcpUrl: string;
  private readonly fetchImpl: typeof fetch;
  private readonly authToken: string | undefined;
  private readonly timeoutMs: number;
  private readonly cacheTtlMs: number;
  private cachedTools: { tools: AgenticToolDefinition[]; timestamp: number } | null = null;
  private mutableSessionId: string | null;

  constructor(config: SuperserverMcpClientConfig = {}) {
    this.mcpUrl = normalizeMcpUrl(config.mcpUrl ?? DEFAULT_MCP_URL);
    this.fetchImpl = resolveFetch(config.fetcher);
    this.authToken = config.authToken;
    this.timeoutMs = config.timeoutMs ?? DEFAULT_MCP_TIMEOUT_MS;
    this.cacheTtlMs = config.cacheTtlMs ?? DEFAULT_TOOL_CACHE_TTL_MS;
    this.mutableSessionId = config.sessionId ?? null;
  }

  get sessionId(): string | null {
    return this.mutableSessionId;
  }

  async preflight(): Promise<McpPreflightResult> {
    const start = Date.now();
    await this.initialize();
    const cachedBeforeList =
      this.cachedTools !== null &&
      Date.now() - this.cachedTools.timestamp < this.cacheTtlMs;
    const tools = await this.listTools();

    return {
      tools,
      sessionId: this.mutableSessionId,
      durationMs: Date.now() - start,
      cached: cachedBeforeList,
    };
  }

  async listTools(options?: {
    forceRefresh?: boolean;
  }): Promise<AgenticToolDefinition[]> {
    if (
      !options?.forceRefresh &&
      this.cachedTools &&
      Date.now() - this.cachedTools.timestamp < this.cacheTtlMs
    ) {
      return this.cachedTools.tools;
    }

    await this.initialize();
    const result = await this.callRpc('tools/list', {});
    const tools = extractMcpTools(result).map(toOpenAIToolDefinition);
    this.cachedTools = { tools, timestamp: Date.now() };
    return tools;
  }

  async callTool(
    name: string,
    args: Record<string, unknown>,
    options?: { toolCallId?: string },
  ): Promise<ToolExecutionResult> {
    const startedAt = Date.now();
    try {
      const result = await this.callRpc('tools/call', {
        name,
        arguments: args,
      });
      const decoded = decodeMcpToolResult(result);
      return {
        toolCallId: options?.toolCallId ?? createId('call'),
        toolName: name,
        args,
        result: decoded.result,
        success: !decoded.error,
        durationMs: Date.now() - startedAt,
        ...(decoded.error ? { error: decoded.error } : {}),
      };
    } catch (error) {
      return {
        toolCallId: options?.toolCallId ?? createId('call'),
        toolName: name,
        args,
        result: null,
        success: false,
        durationMs: Date.now() - startedAt,
        error: error instanceof Error ? error.message : String(error),
      };
    }
  }

  private async initialize(): Promise<void> {
    if (this.mutableSessionId) return;
    await this.callRpc('initialize', {
      protocolVersion: '2024-11-05',
      capabilities: {},
      clientInfo: {
        name: 'distributed-inference-host',
        version: '0.1.0',
      },
    });
  }

  private async callRpc(
    method: string,
    params: Record<string, unknown>,
  ): Promise<unknown> {
    const envelope: RpcEnvelope = {
      jsonrpc: '2.0',
      id: createId('rpc'),
      method,
      params,
    };

    const headers = new Headers({
      'content-type': 'application/json',
    });
    if (this.authToken) {
      headers.set('authorization', `Bearer ${this.authToken}`);
    }
    if (this.mutableSessionId) {
      headers.set(MCP_SESSION_HEADER, this.mutableSessionId);
    }

    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), this.timeoutMs);
    try {
      const response = await this.fetchImpl(this.mcpUrl, {
        method: 'POST',
        headers,
        body: JSON.stringify(envelope),
        signal: controller.signal,
      });
      const nextSessionId = response.headers.get(MCP_SESSION_HEADER);
      if (nextSessionId) {
        this.mutableSessionId = nextSessionId;
      }
      if (!response.ok) {
        throw new Error(`MCP request failed with HTTP ${response.status}`);
      }

      const rpc = (await response.json()) as RpcResponse;
      if (rpc.error) {
        const code = rpc.error.code ?? -32603;
        const message = rpc.error.message ?? 'MCP request failed';
        throw new Error(`MCP error ${code}: ${message}`);
      }
      return rpc.result;
    } finally {
      clearTimeout(timeout);
    }
  }
}

/**
 * Creates the Superserver Mcp Client.
 */
export function createSuperserverMcpClient(
  config: SuperserverMcpClientConfig = {},
): SuperserverMcpClient {
  return new SuperserverMcpClient(config);
}

/**
 * Handles the gnosis run Agentic Chat Completion workflow.
 */
export async function runAgenticChatCompletion(
  request: AgenticChatRequest,
  runtime: AgenticChatRuntime,
): Promise<AgenticChatCompletionResponse> {
  const now = runtime.now ?? (() => Date.now());
  const randomId = runtime.randomId ?? createId;
  const nativeFamily = runtime.nativeToolFamily ?? null;
  const trace: AgenticTraceEntry[] = [];
  const startedAt = now();
  const created = Math.floor(startedAt / 1000);
  const responseId = randomId('chatcmpl');
  const toolChoice = request.tool_choice ?? 'auto';
  const autoTools = request.auto_tools !== false;
  const executeTools = request.execute_tools !== false;
  const evaluateIntent = request.evaluate_intent !== false;
  const maxToolRounds = Math.max(
    1,
    Math.floor(request.max_tool_rounds ?? DEFAULT_MAX_TOOL_ROUNDS),
  );
  const intentThreshold =
    request.intent_confidence_threshold ?? DEFAULT_INTENT_CONFIDENCE;
  const maxNudges = request.max_continuation_nudges ?? DEFAULT_MAX_NUDGES;
  const maxToolsInPrompt = request.max_tools_in_prompt;
  const metacognitionEnabled = request.metacognition?.enabled !== false;
  const reviewToolActivations =
    metacognitionEnabled &&
    request.metacognition?.reviewToolActivations !== false;
  const validateToolResults =
    metacognitionEnabled &&
    request.metacognition?.validateToolResults !== false;
  const maxValidationRetries = Math.max(
    1,
    request.metacognition?.maxValidationRetries ?? 3,
  );

  let promptTokens = 0;
  let completionTokens = 0;
  let modelUsed = request.model;
  const withOathMcp = composeAgenticMcpClients(
    runtime.mcp,
    request.thothGateTelemetry
      ? createThothOathSelfManagementMcpClient({
          now,
          randomId,
          ledgerStore: runtime.thothOathLedgerStore,
        })
      : undefined,
  );
  // SENSE-OF-PLACE tools (thoth/place.*): the headless living-city sim, same
  // local-client pattern as the oath tools. The request observer seeds the
  // default city so the model can sample/recommend without an explicit load.
  const placeToolsLive = Boolean(
    request.thothPlaceTools && runtime.thothPlaceSimHost,
  );
  const effectiveMcp = composeAgenticMcpClients(
    withOathMcp,
    placeToolsLive
      ? createThothPlaceMcpClient({
          host: runtime.thothPlaceSimHost!,
          observer: request.thothPlaceObserver ?? null,
          now,
          randomId,
        })
      : undefined,
  );
  let toolPreflight: AgenticChatCompletionResponse['agentic']['toolPreflight'] =
    {
      enabled: autoTools && toolChoice !== 'none',
      toolCount: 0,
      visibleToolCount: 0,
      cached: false,
      durationMs: 0,
    };
  const toolsUsed: string[] = [];
  const economy = new AgentEconomy();
  const currentMessages = request.messages.map((message) => ({ ...message }));
  const originalUserMessage = getLastUserMessage(currentMessages);
  if (request.thothGateTelemetry) {
    injectThothGateTelemetryPrompt(currentMessages, request.thothGateTelemetry);
    trace.push({
      layer: 'C0',
      message: 'Thoth gate telemetry threaded into prompt/nudge/action',
      data: {
        gateCount: request.thothGateTelemetry.scores.length,
        totalPressure: roundScore(
          thothGateTelemetryTotalPressure(request.thothGateTelemetry),
        ),
        actionHint: thothGateTelemetryActionHint(request.thothGateTelemetry),
      },
    });
  }
  if (request.thothNpcHalogram) {
    injectThothNpcHalogramPrompt(currentMessages, request.thothNpcHalogram);
    trace.push({
      layer: 'C1',
      message: 'Shared-ui Halogram NPC projection threaded into prompt',
      data: {
        npcId: request.thothNpcHalogram.npcId,
        projectionSource: request.thothNpcHalogram.projectionSource,
        calibratedConfidences:
          request.thothNpcHalogram.calibratedConfidenceCount,
        coherence: roundScore(request.thothNpcHalogram.overallCoherence),
        projectionConfidence: roundScore(
          request.thothNpcHalogram.confidenceInProjection,
        ),
      },
    });
  }
  if (request.thothNearbyBusinesses) {
    // SLIM GATE: when the thoth/place.* tools are live AND tools will actually
    // be injected into the prompt (small-model runs set max_tools_in_prompt 0
    // and never see tools), the 14-line business dump collapses to a 2-line
    // pointer — the model pulls venue detail on demand instead of paying for
    // it every request. Otherwise the full stuffed render is preserved
    // verbatim (existing clients + tiny models unregressed).
    const slimPointer = placeToolsLive && maxToolsInPrompt !== 0;
    if (slimPointer) {
      currentMessages.unshift({
        role: 'system',
        content: renderThothPlacePointerPromptContext(
          request.thothNearbyBusinesses,
        ),
      });
    } else {
      injectThothNearbyBusinessesPrompt(
        currentMessages,
        request.thothNearbyBusinesses,
      );
    }
    trace.push({
      layer: 'C0',
      message: 'Thoth nearby-businesses sense of place threaded into prompt',
      data: {
        cityName: request.thothNearbyBusinesses.cityName,
        businessCount: request.thothNearbyBusinesses.businessCount,
        source: request.thothNearbyBusinesses.source,
        slimmed: slimPointer,
      },
    });
  }

  let allTools: AgenticToolDefinition[] = (request.tools ?? []).map((tool) => ({
    ...tool,
    source: tool.source ?? 'provided',
  }));

  if (autoTools && toolChoice !== 'none' && request.tools === undefined) {
    const preflight = await preflightTools(effectiveMcp);
    allTools = preflight.tools;
    toolPreflight = {
      enabled: true,
      toolCount: preflight.tools.length,
      visibleToolCount: 0,
      cached: preflight.cached,
      durationMs: preflight.durationMs,
      sessionId: preflight.sessionId,
    };
    trace.push({
      layer: 'preflight',
      message: 'Superserver MCP tools preflighted',
      data: {
        toolCount: preflight.tools.length,
        cached: preflight.cached,
        sessionId: preflight.sessionId ?? null,
      },
    });
  }

  const activation =
    originalUserMessage && allTools.length > 0 && toolChoice !== 'none'
      ? activateTools(originalUserMessage, allTools)
      : emptyActivation();
  if (activation.candidates.length > 0) {
    trace.push({
      layer: 'C0',
      message: 'Tool activation candidates identified',
      data: {
        candidates: activation.candidates.slice(0, 5).map((candidate) => ({
          toolName: candidate.toolName,
          score: roundScore(candidate.score),
          matches: candidate.matchedPatterns,
        })),
      },
    });
    trace.push({
      layer: 'C1',
      message: 'Activation entropy computed',
      data: {
        entropy: roundScore(activation.entropy),
        autoActivate: activation.autoActivate,
      },
    });
  }

  let visibleTools = selectVisibleTools(allTools, activation, toolChoice);
  toolPreflight = {
    ...toolPreflight,
    visibleToolCount: visibleTools.length,
  };
  // Skip tool injection entirely when max_tools_in_prompt is 0 — the model is
  // too small to follow structured tool instructions without generating garbage.
  if (visibleTools.length > 0 && toolChoice !== 'none' && maxToolsInPrompt !== 0) {
    injectToolDeclaration(currentMessages, visibleTools, maxToolsInPrompt, nativeFamily);
  }

  if (
    activation.autoActivate &&
    activation.topCandidate &&
    executeTools &&
    effectiveMcp
  ) {
    const review = reviewToolActivations
      ? await reviewActivation(
          request,
          runtime,
          activation.topCandidate,
          originalUserMessage,
          currentMessages,
          request.thothGateTelemetry,
        )
      : { action: 'accept' as const };
    trace.push({
      layer: 'C2',
      message: 'Heuristic activation reviewed',
      data: {
        toolName: activation.topCandidate.toolName,
        action: review.action,
      },
    });

    if (review.action !== 'cancel') {
      const finalArgs =
        review.action === 'modify' && review.modifiedArgs
          ? review.modifiedArgs
          : await resolveToolArgs(
              request,
              runtime,
              activation.topCandidate,
              originalUserMessage,
              currentMessages,
            );
      const toolCall = buildToolCall(
        activation.topCandidate.toolName,
        finalArgs,
        randomId,
      );
      const result = await executeToolWithValidation({
        toolCall,
        mcp: effectiveMcp,
        runtime,
        request,
        currentMessages,
        originalUserMessage,
        validateToolResults,
        maxValidationRetries,
        trace,
      });
      appendToolExchange(currentMessages, toolCall, result);
      toolsUsed.push(result.toolName);
      if (result.success) {
        economy.recordSuccess(
          result.toolName,
          classifyToolRisk(result.toolName),
          result.args,
          serializeToolResult(result),
        );
      } else {
        economy.recordFailure(classifyToolRisk(result.toolName));
      }
    }
  } else if (activation.candidates.length > 0) {
    injectNarrowedToolContext(currentMessages, activation.candidates);
  }

  let round = 0;
  let nudgeCount = 0;
  let finalResponseText = '';
  let finalToolCalls: AgenticToolCall[] | undefined;
  let maxToolRoundsReached = false;
  let lastToolResults: ToolExecutionResult[] = [];

  while (round < maxToolRounds) {
    round += 1;
    trace.push({
      layer: 'loop',
      message: 'Agentic chat round started',
      data: { round, maxToolRounds },
    });

    if (!economy.isAtRest()) {
      // Nothing to report before the first tool call/result — injecting the
      // block only once there's real state avoids a small/weak model latching
      // onto static boilerplate and echoing it back instead of answering or
      // calling a tool (same reasoning as Moonshine's `is_at_rest()` gate).
      currentMessages.push({
        role: 'system',
        content: economyStatusBlock(economy),
      });
    }

    const generation = await runtime.infer({
      model: request.model,
      messages: currentMessages,
      temperature: request.temperature,
      maxTokens: request.max_tokens,
      responseFormat: request.response_format,
      round,
      purpose: 'chat',
      tools: visibleTools.length > 0 ? visibleTools : undefined,
    });
    promptTokens += generation.promptTokens ?? estimateTokensFromMessages(currentMessages);
    completionTokens +=
      generation.completionTokens ?? estimateTokens(generation.content);
    modelUsed = generation.model ?? modelUsed;

    const validToolNames = new Set(visibleTools.map((tool) => tool.function.name));
    const parsed =
      visibleTools.length > 0 && request.response_format?.type !== 'json_object'
        ? parseModelOutput(generation.content, validToolNames, nativeFamily)
        : { text: generation.content.trim(), toolCalls: null };

    if (!parsed.toolCalls || !executeTools) {
      if (
        evaluateIntent &&
        executeTools &&
        originalUserMessage &&
        nudgeCount < maxNudges
      ) {
        const evaluation = await evaluateIntentSatisfaction(
          request,
          runtime,
          originalUserMessage,
          currentMessages,
          lastToolResults,
        );
        trace.push({
          layer: 'intent',
          message: 'Intent satisfaction evaluated',
          data: {
            satisfied: evaluation.satisfied,
            confidence: roundScore(evaluation.confidence),
            reasoning: evaluation.reasoning,
          },
        });
        if (!evaluation.satisfied && evaluation.confidence >= intentThreshold) {
          nudgeCount += 1;
          if (parsed.text) {
            currentMessages.push({
              role: 'assistant',
              content: parsed.text,
            });
          }
          currentMessages.push({
            role: 'system',
            content: continuationNudgeWithTelemetry(
              evaluation,
              request.thothGateTelemetry,
            ),
          });
          continue;
        }
      }

      finalResponseText = parsed.text;
      finalToolCalls = !executeTools ? parsed.toolCalls ?? undefined : undefined;
      economy.awardFinisherIfCiting(finalResponseText);
      break;
    }

    currentMessages.push({
      role: 'assistant',
      content: parsed.text.length > 0 ? parsed.text : null,
      tool_calls: parsed.toolCalls,
    });

    lastToolResults = await Promise.all(
      parsed.toolCalls.map((toolCall) =>
        executeToolWithValidation({
          toolCall,
          mcp: effectiveMcp,
          runtime,
          request,
          currentMessages,
          originalUserMessage,
          validateToolResults,
          maxValidationRetries,
          trace,
        }),
      ),
    );

    for (const result of lastToolResults) {
      toolsUsed.push(result.toolName);
      currentMessages.push({
        role: 'tool',
        tool_call_id: result.toolCallId,
        name: result.toolName,
        content: serializeToolResult(result),
      });
      if (result.success) {
        economy.recordSuccess(
          result.toolName,
          classifyToolRisk(result.toolName),
          result.args,
          serializeToolResult(result),
        );
      } else {
        economy.recordFailure(classifyToolRisk(result.toolName));
      }
    }
  }

  if (!finalResponseText && round >= maxToolRounds) {
    maxToolRoundsReached = true;
    const generation = await runtime.infer({
      model: request.model,
      messages: currentMessages,
      temperature: request.temperature,
      maxTokens: request.max_tokens,
      responseFormat: request.response_format,
      round: round + 1,
      purpose: 'chat',
      tools: visibleTools.length > 0 ? visibleTools : undefined,
    });
    promptTokens += generation.promptTokens ?? estimateTokensFromMessages(currentMessages);
    completionTokens +=
      generation.completionTokens ?? estimateTokens(generation.content);
    modelUsed = generation.model ?? modelUsed;
    finalResponseText = parseModelOutput(
      generation.content,
      new Set(visibleTools.map((tool) => tool.function.name)),
      nativeFamily,
    ).text;
    economy.awardFinisherIfCiting(finalResponseText);
  }

  const uniqueToolsUsed = Array.from(new Set(toolsUsed));
  return {
    id: responseId,
    object: 'chat.completion',
    created,
    model: modelUsed,
    choices: [
      {
        index: 0,
        message: {
          role: 'assistant',
          content: finalToolCalls ? null : finalResponseText,
          ...(finalToolCalls ? { tool_calls: finalToolCalls } : {}),
        },
        finish_reason: finalToolCalls ? 'tool_calls' : 'stop',
      },
    ],
    usage: {
      prompt_tokens: promptTokens,
      completion_tokens: completionTokens,
      total_tokens: promptTokens + completionTokens,
    },
    ...(uniqueToolsUsed.length > 0 ? { tools_used: uniqueToolsUsed } : {}),
    agentic: {
      rounds: round,
      toolPreflight,
      metacognitionTrace: trace,
      nudges: nudgeCount,
      maxToolRoundsReached,
      economy: economy.summary(),
    },
  };
}

/**
 * Handles the gnosis render Messages For Prompt workflow.
 */
export function renderMessagesForPrompt(messages: AgenticChatMessage[]): string {
  return messages
    .map((message) => {
      if (message.role === 'tool') {
        return `tool:${message.name ?? 'unknown'}: ${message.content ?? ''}`;
      }
      if (message.tool_calls && message.tool_calls.length > 0) {
        const calls = message.tool_calls
          .map((call) => `${call.function.name}(${call.function.arguments})`)
          .join(', ');
        return `${message.role}: [tool_calls: ${calls}]${
          message.content ? `\n${message.content}` : ''
        }`;
      }
      return `${message.role}: ${message.content ?? ''}`;
    })
    .join('\n\n');
}

const DISCOVERY_TOOL_NAMES = new Set(['network/search', 'network/directory', 'network/docs']);

function selectDiscoveryTools(tools: AgenticToolDefinition[]): AgenticToolDefinition[] {
  const found = tools.filter((t) => DISCOVERY_TOOL_NAMES.has(t.function.name));
  return found.length > 0 ? found : tools.slice(0, 3);
}

/**
 * Formats the Tools For Prompt.
 */
export function formatToolsForPrompt(tools: AgenticToolDefinition[]): string {
  if (tools.length === 0) return '';

  const toolDescriptions = tools
    .map((tool) => {
      const params = tool.function.parameters?.properties
        ? Object.entries(tool.function.parameters.properties)
            .map(([key, value]) => {
              const field = isRecord(value) ? value : {};
              const desc =
                typeof field.description === 'string'
                  ? field.description
                  : typeof field.type === 'string'
                  ? field.type
                  : 'unknown';
              const required = tool.function.parameters?.required?.includes(key)
                ? ' (required)'
                : '';
              return `  - ${key}: ${desc}${required}`;
            })
            .join('\n')
        : '  (no parameters)';

      return `### ${tool.function.name}
${tool.function.description ?? 'No description provided.'}
Parameters:
${params}`;
    })
    .join('\n\n');

  return `## TOOL CALLING INSTRUCTIONS

You have access to ${tools.length} tools from the ForkJoin MCP superserver.
Use a tool when it materially advances the user's request.

Use this exact format:

\`\`\`tool_call
{"name": "tool_name", "arguments": {"param": "value"}}
\`\`\`

Rules:
1. The tool name must exactly match one of the available tools.
2. Arguments must be a JSON object.
3. You may call multiple tools, one fenced tool_call block per tool.
4. If no tool is useful, answer normally.

Available tools:

${toolDescriptions}
`;
}

/**
 * Formats the Tools For Prompt Discovery.
 */
export function formatToolsForPromptDiscovery(
  discoveryTools: AgenticToolDefinition[],
  _totalToolCount: number,
): string {
  if (discoveryTools.length === 0) return '';
  const toolNames = discoveryTools.map((t) => t.function.name).join(', ');
  return `You can search for tools using network/search. Tools available for discovery: ${toolNames}. Use a tool only if it directly helps the user.`;
}

function normalizeMcpUrl(value: string): string {
  const url = new URL(value);
  if (url.pathname === '/' || url.pathname.length === 0) {
    url.pathname = '/mcp';
  }
  return url.toString();
}

function resolveFetch(
  fetcher?: Pick<typeof globalThis, 'fetch'>['fetch'] | { fetch: typeof fetch },
): typeof fetch {
  if (!fetcher) {
    if (typeof fetch !== 'function') {
      throw new Error('No fetch implementation available for SuperserverMcpClient');
    }
    return fetch.bind(globalThis);
  }
  if (typeof fetcher === 'function') {
    return fetcher;
  }
  return fetcher.fetch.bind(fetcher);
}

async function preflightTools(
  mcp: AgenticMcpClient | undefined,
): Promise<McpPreflightResult> {
  const startedAt = Date.now();
  if (!mcp) {
    return { tools: [], durationMs: 0, cached: false };
  }
  if (mcp.preflight) {
    return mcp.preflight();
  }
  const tools = await mcp.listTools();
  return {
    tools,
    sessionId: mcp.sessionId,
    durationMs: Date.now() - startedAt,
    cached: false,
  };
}

function composeAgenticMcpClients(
  primary: AgenticMcpClient | undefined,
  secondary: AgenticMcpClient | undefined,
): AgenticMcpClient | undefined {
  if (!primary) return secondary;
  if (!secondary) return primary;
  return {
    sessionId: primary.sessionId ?? secondary.sessionId,
    async preflight() {
      const [left, right] = await Promise.all([
        preflightTools(primary),
        preflightTools(secondary),
      ]);
      return {
        tools: mergeToolDefinitions(left.tools, right.tools),
        sessionId: [left.sessionId, right.sessionId].filter(Boolean).join('|') || null,
        durationMs: left.durationMs + right.durationMs,
        cached: left.cached && right.cached,
      };
    },
    async listTools(options?: { forceRefresh?: boolean }) {
      const [left, right] = await Promise.all([
        primary.listTools(options),
        secondary.listTools(options),
      ]);
      return mergeToolDefinitions(left, right);
    },
    async callTool(
      name: string,
      args: Record<string, unknown>,
      options?: { toolCallId?: string },
    ) {
      const secondaryTools = await secondary.listTools();
      if (secondaryTools.some((tool) => tool.function.name === name)) {
        return secondary.callTool(name, args, options);
      }
      return primary.callTool(name, args, options);
    },
  };
}

function mergeToolDefinitions(
  primary: AgenticToolDefinition[],
  secondary: AgenticToolDefinition[],
): AgenticToolDefinition[] {
  const byName = new Map<string, AgenticToolDefinition>();
  for (const tool of primary) byName.set(tool.function.name, tool);
  for (const tool of secondary) byName.set(tool.function.name, tool);
  return Array.from(byName.values());
}

function extractMcpTools(result: unknown): McpToolWire[] {
  if (!isRecord(result)) return [];
  const tools = result.tools;
  if (!Array.isArray(tools)) return [];
  return tools.filter(isMcpToolWire);
}

function isMcpToolWire(value: unknown): value is McpToolWire {
  return isRecord(value) && typeof value.name === 'string';
}

function toOpenAIToolDefinition(tool: McpToolWire): AgenticToolDefinition {
  const metadata = Object.fromEntries(
    Object.entries(tool).filter(
      ([key]) =>
        key !== 'name' &&
        key !== 'description' &&
        key !== 'inputSchema' &&
        key !== 'activation',
    ),
  );
  return {
    type: 'function',
    function: {
      name: tool.name,
      ...(tool.description ? { description: tool.description } : {}),
      parameters: {
        type: 'object',
        properties: tool.inputSchema?.properties ?? {},
        ...(tool.inputSchema?.required ? { required: tool.inputSchema.required } : {}),
      },
    },
    ...(tool.activation ? { activation: tool.activation } : {}),
    source: 'superserver',
    ...(Object.keys(metadata).length > 0 ? { metadata } : {}),
  };
}

function decodeMcpToolResult(result: unknown): {
  result: unknown;
  error?: string;
} {
  if (!isRecord(result)) return { result };
  const content = result.content;
  const isError = result.isError === true;
  if (Array.isArray(content)) {
    const text = content
      .filter(isMcpContentBlock)
      .filter((block) => block.type === 'text' && typeof block.text === 'string')
      .map((block) => block.text)
      .join('\n');
    const parsed = parseJsonIfPossible(text);
    return {
      result: parsed,
      ...(isError ? { error: typeof parsed === 'string' ? parsed : text } : {}),
    };
  }
  return {
    result,
    ...(isError ? { error: 'MCP tool returned isError' } : {}),
  };
}

function isMcpContentBlock(value: unknown): value is McpContentBlock {
  return isRecord(value) && typeof value.type === 'string';
}

function parseJsonIfPossible(value: string): unknown {
  const trimmed = value.trim();
  if (!trimmed) return '';
  try {
    return JSON.parse(trimmed);
  } catch {
    return value;
  }
}

function emptyActivation(): ActivationResult {
  return {
    candidates: [],
    topCandidate: null,
    entropy: 0,
    autoActivate: false,
  };
}

/**
 * Document-frequency ratio above which a token is treated as corpus noise
 * rather than a discriminating signal — calibrated against the real live
 * catalog (8490 tools): the ~23 tokens above this line are all boilerplate
 * ("ledger" 81%, "thm" 81%, "verification" 15.7%, "mcp" 14.25%, domain words
 * like "gnosis"/"math"/"topology" at 7-15%), while every genuinely useful
 * word measured ("search" 0.73%, "tool" 1.13%, "network" 0.68%, "local"
 * 0.85%, "file" 0.14%, "registry" 0.04%) sits comfortably under 2%.
 */
export const COMMON_TOKEN_DOCUMENT_FREQUENCY_RATIO = 0.05;

/**
 * Which name/description tokens are common enough across `tools` to be
 * corpus-level noise (see `COMMON_TOKEN_DOCUMENT_FREQUENCY_RATIO`) — the
 * same idea as the fixed `STOP_WORDS` list above, but derived from the real
 * catalog instead of hand-picked: a word every tool mentions doesn't
 * discriminate, no matter how relevant it looks in isolation. Pure and
 * synchronous, so it's cheap to unit test and cheap to memoize (see
 * `getCommonTokensForToolList`) against the real tool-list array, which only
 * changes on a superserver preflight cache refresh.
 */
export function computeCommonTokens(
  tools: AgenticToolDefinition[],
  ratio: number = COMMON_TOKEN_DOCUMENT_FREQUENCY_RATIO,
): Set<string> {
  const documentFrequency = new Map<string, number>();
  for (const tool of tools) {
    const tokens = new Set([
      ...tokenizeToolText(tool.function.name),
      ...tokenizeToolText(tool.function.description ?? ''),
    ]);
    for (const token of tokens) {
      documentFrequency.set(token, (documentFrequency.get(token) ?? 0) + 1);
    }
  }
  const threshold = tools.length * ratio;
  const common = new Set<string>();
  for (const [token, count] of documentFrequency) {
    if (count >= threshold) common.add(token);
  }
  return common;
}

// Keyed by the tools array's own reference: `SuperserverMcpClient.listTools()`
// returns the same array instance while its TTL cache is warm and only a new
// instance on refresh, so this never needs explicit invalidation — the old
// array becomes unreachable and the WeakMap drops it naturally.
const commonTokensByToolList = new WeakMap<AgenticToolDefinition[], Set<string>>();

/**
 * Handles the gnosis get Common Tokens For Tool List workflow.
 */
export function getCommonTokensForToolList(tools: AgenticToolDefinition[]): Set<string> {
  let cached = commonTokensByToolList.get(tools);
  if (!cached) {
    cached = computeCommonTokens(tools);
    commonTokensByToolList.set(tools, cached);
  }
  return cached;
}

function activateTools(
  userMessage: string,
  tools: AgenticToolDefinition[],
): ActivationResult {
  const input = userMessage.toLowerCase();
  const commonTokens = getCommonTokensForToolList(tools);
  const candidates: ToolCandidate[] = [];

  for (const tool of tools) {
    const activation = tool.activation;
    const toolName = tool.function.name;
    const matchedPatterns: string[] = [];
    let bestScore = 0;

    if (activation?.patterns) {
      for (const pattern of activation.patterns) {
        if (matchesPattern(input, pattern)) {
          bestScore = Math.max(bestScore, 0.85);
          matchedPatterns.push(pattern);
        }
      }
    }

    if (activation?.keywords) {
      for (const keyword of activation.keywords) {
        if (input.includes(keyword.toLowerCase())) {
          bestScore = Math.max(bestScore, 0.9);
          matchedPatterns.push(`kw:${keyword}`);
        }
      }
    }

    if (!activation) {
      const derived = scoreDerivedToolMatch(input, tool, commonTokens);
      if (derived.score > 0) {
        bestScore = Math.max(bestScore, derived.score);
        matchedPatterns.push(...derived.matches);
      }
    }

    if (bestScore >= 0.3 && matchedPatterns.length > 0) {
      const multiMatchBoost = Math.min(matchedPatterns.length * 0.03, 0.1);
      candidates.push({
        toolName,
        score: Math.min(bestScore + multiMatchBoost, 1),
        matchedPatterns,
        suggestedArgs: extractArguments(toolName, userMessage, activation),
        ...(activation?.onlyRules ? { onlyRules: true } : {}),
        ...(activation?.skipReview ? { skipReview: true } : {}),
        ...(activation?.agentResponse !== undefined
          ? { agentResponse: activation.agentResponse }
          : {}),
        ...(activation?.tier !== undefined ? { tier: activation.tier } : {}),
      });
    }
  }

  candidates.sort((left, right) => {
    const tierDiff = (left.tier ?? 3) - (right.tier ?? 3);
    return tierDiff === 0 ? right.score - left.score : tierDiff;
  });

  const entropy = calculateEntropy(candidates.map((candidate) => candidate.score));
  const topCandidate = candidates[0] ?? null;
  const topTool = topCandidate
    ? tools.find((tool) => tool.function.name === topCandidate.toolName)
    : undefined;
  const threshold =
    topTool?.activation?.threshold ?? DEFAULT_AUTO_ACTIVATE_THRESHOLD;
  const autoActivate =
    topCandidate !== null &&
    topCandidate.score >= threshold &&
    entropy <= AUTO_ACTIVATE_MAX_ENTROPY &&
    (candidates.length === 1 || topCandidate.score - candidates[1]!.score > 0.15);

  return { candidates, topCandidate, entropy, autoActivate };
}

function matchesPattern(input: string, pattern: string): boolean {
  try {
    return new RegExp(pattern, 'i').test(input);
  } catch {
    return input.includes(pattern.toLowerCase());
  }
}

function scoreDerivedToolMatch(
  input: string,
  tool: AgenticToolDefinition,
  commonTokens: Set<string>,
): { score: number; matches: string[] } {
  // Dedupe: a token repeating many times in a tool's own name/description
  // (e.g. boilerplate text mentioning "MCP" five times) must not inflate the
  // match count on its own — only DISTINCT signal words matched should count
  // toward the multi-match score bumps below and `activateTools`'s
  // `multiMatchBoost`. Without this, any query sharing one common,
  // repeated-in-boilerplate word ties a flood of unrelated tools together at
  // an inflated score, drowning out the tool that's actually relevant.
  //
  // `commonTokens` layers on a second, corpus-frequency-aware filter: even a
  // DISTINCT (non-repeated) match on a word nearly every tool's name/
  // description contains ("verification", "mcp") is not a discriminating
  // signal either — see `computeCommonTokens`.
  const nameTokens = new Set(
    [...tokenizeToolText(tool.function.name)].filter((t) => !commonTokens.has(t)),
  );
  const descTokens = new Set(
    [...tokenizeToolText(tool.function.description ?? '')]
      .slice(0, 24)
      .filter((t) => !commonTokens.has(t)),
  );
  let score = 0;
  const matches: string[] = [];

  for (const token of nameTokens) {
    if (input.includes(token)) {
      score = Math.max(score, 0.55);
      matches.push(`name:${token}`);
    }
  }

  for (const token of descTokens) {
    if (input.includes(token)) {
      score = Math.max(score, 0.4);
      matches.push(`desc:${token}`);
    }
  }

  if (matches.length >= 3) {
    score = Math.max(score, 0.65);
  }

  return { score, matches };
}

/**
 * Converts input into tokenize Tool Text.
 */
export function tokenizeToolText(value: string): string[] {
  return value
    .toLowerCase()
    .split(/[^a-z0-9]+/u)
    .filter((part) => part.length >= 3 && !STOP_WORDS.has(part));
}

function calculateEntropy(scores: number[]): number {
  if (scores.length <= 1) return 0;
  const total = scores.reduce((sum, score) => sum + score, 0);
  if (total === 0) return 0;
  const maxEntropy = Math.log2(scores.length);
  const entropy = scores.reduce((sum, score) => {
    const p = score / total;
    return p > 0 ? sum - p * Math.log2(p) : sum;
  }, 0);
  return maxEntropy > 0 ? entropy / maxEntropy : 0;
}

function extractArguments(
  toolName: string,
  userMessage: string,
  activation?: ToolActivationMetadata,
): Record<string, unknown> {
  const args: Record<string, unknown> = {};
  if (activation?.argExtractors) {
    for (const [argName, pattern] of Object.entries(activation.argExtractors)) {
      try {
        const match = userMessage.match(new RegExp(pattern, 'i'));
        if (match?.[1]) args[argName] = match[1];
      } catch {
        // Ignore invalid tool-provided extractor patterns.
      }
    }
  }
  if (Object.keys(args).length > 0) return args;

  if (
    toolName.endsWith('/search') ||
    toolName.endsWith('_search') ||
    toolName.includes('search')
  ) {
    return { query: userMessage };
  }
  if (
    toolName.includes('lint') ||
    toolName.includes('compile') ||
    toolName.includes('verify') ||
    toolName.includes('analyze')
  ) {
    const quoted = userMessage.match(/```[\s\S]*?\n([\s\S]*?)```/u);
    return quoted?.[1] ? { source: quoted[1] } : {};
  }
  return {};
}

function selectVisibleTools(
  tools: AgenticToolDefinition[],
  activation: ActivationResult,
  toolChoice: AgenticToolChoice,
): AgenticToolDefinition[] {
  if (toolChoice === 'none') return [];
  if (typeof toolChoice === 'object') {
    return tools.filter(
      (tool) => tool.function.name === toolChoice.function.name && !tool.activation?.onlyRules,
    );
  }
  if (tools.length <= DEFAULT_MAX_VISIBLE_TOOLS) {
    return tools.filter((tool) => !tool.activation?.onlyRules);
  }

  const selected = new Map<string, AgenticToolDefinition>();
  const add = (tool: AgenticToolDefinition | undefined): void => {
    if (!tool || tool.activation?.onlyRules) return;
    if (selected.size >= DEFAULT_MAX_VISIBLE_TOOLS) return;
    selected.set(tool.function.name, tool);
  };

  for (const tool of tools) {
    if (PINNED_GATEWAY_TOOLS.has(tool.function.name)) add(tool);
  }
  for (const candidate of activation.candidates) {
    add(tools.find((tool) => tool.function.name === candidate.toolName));
  }
  for (const tool of tools) {
    add(tool);
  }

  return Array.from(selected.values());
}

function spliceSystemPrompt(messages: AgenticChatMessage[], text: string): void {
  const systemIndex = messages.findIndex(
    (message) => message.role === 'system' || message.role === 'developer',
  );
  if (systemIndex >= 0) {
    const existing = messages[systemIndex]!;
    existing.content = `${text}\n\n${existing.content ?? ''}`.trim();
    return;
  }
  messages.unshift({ role: 'system', content: text });
}

function injectToolPrompt(
  messages: AgenticChatMessage[],
  tools: AgenticToolDefinition[],
  maxToolsInPrompt?: number,
): void {
  const useDiscovery =
    maxToolsInPrompt !== undefined && tools.length > maxToolsInPrompt;
  const toolPrompt = useDiscovery
    ? formatToolsForPromptDiscovery(selectDiscoveryTools(tools), tools.length)
    : formatToolsForPrompt(tools);
  spliceSystemPrompt(messages, toolPrompt);
}

/** Splices a native family's tool declaration into the message array when it
 * fits as system-message content (Qwen, Gemma4). Mistral's `[AVAILABLE_TOOLS]`
 * segment is NOT handled here — it's threaded via `TextGenerationRequest.tools`
 * instead, since it must render before the first `[INST]` turn, outside any
 * message's content. Falls back to the generic fenced-block prompt when
 * `maxToolsInPrompt` forces discovery mode (native formats have no discovery
 * variant defined) or when no native family is active. Returns the resolved
 * family for the parse-side dispatcher to reuse (avoids re-resolving).
 */
function injectToolDeclaration(
  messages: AgenticChatMessage[],
  tools: AgenticToolDefinition[],
  maxToolsInPrompt: number | undefined,
  nativeFamily: NativeToolFamily | null,
): void {
  const useDiscovery =
    maxToolsInPrompt !== undefined && tools.length > maxToolsInPrompt;
  if (nativeFamily && !useDiscovery) {
    const declaration = formatNativeToolsDeclaration(nativeFamily, tools);
    if (declaration.systemAddition) {
      spliceSystemPrompt(messages, declaration.systemAddition);
    }
    return;
  }
  injectToolPrompt(messages, tools, maxToolsInPrompt);
}

/** Parses model output for a tool call, preferring the native family's own
 * parser when active and falling back to the generic fenced-block/bare-JSON
 * parser whenever the native parser finds no family-specific markers at all
 * (capability off, unsupported endpoint, or the model free-texted anyway) —
 * so requests/models without native support are completely unaffected. */
function parseModelOutput(
  output: string,
  validToolNames: Set<string>,
  nativeFamily: NativeToolFamily | null,
): { text: string; toolCalls: AgenticToolCall[] | null } {
  if (nativeFamily) {
    const native = parseNativeToolCalls(nativeFamily, output, validToolNames);
    if (native) return native;
  }
  return parseToolCalls(output, validToolNames);
}

function injectNarrowedToolContext(
  messages: AgenticChatMessage[],
  candidates: ToolCandidate[],
): void {
  const visibleCandidates = candidates.filter((candidate) => !candidate.onlyRules);
  if (visibleCandidates.length === 0) return;
  const lines = visibleCandidates
    .slice(0, 8)
    .map(
      (candidate) =>
        `- ${candidate.toolName} (${Math.round(candidate.score * 100)}% match)`,
    )
    .join('\n');
  messages.unshift({
    role: 'system',
    content: `Relevant preflighted MCP tools for this request:\n${lines}`,
  });
}

function injectThothGateTelemetryPrompt(
  messages: AgenticChatMessage[],
  metrics: ThothGateTelemetryMetrics,
): void {
  messages.unshift({
    role: 'system',
    content: renderThothGateTelemetryPromptContext(metrics),
  });
}

function injectThothNpcHalogramPrompt(
  messages: AgenticChatMessage[],
  metrics: ThothNpcHalogramMetrics,
): void {
  messages.unshift({
    role: 'system',
    content: renderThothNpcHalogramPromptContext(metrics),
  });
}

function injectThothNearbyBusinessesPrompt(
  messages: AgenticChatMessage[],
  metrics: ThothNearbyBusinessesMetrics,
): void {
  messages.unshift({
    role: 'system',
    content: renderThothNearbyBusinessesPromptContext(metrics),
  });
}

async function reviewActivation(
  request: AgenticChatRequest,
  runtime: AgenticChatRuntime,
  candidate: ToolCandidate,
  userMessage: string,
  currentMessages: AgenticChatMessage[],
  thothGateTelemetry?: ThothGateTelemetryMetrics,
): Promise<
  | { action: 'accept' }
  | { action: 'cancel'; reason?: string }
  | { action: 'modify'; modifiedArgs: Record<string, unknown> }
> {
  if (candidate.skipReview) return { action: 'accept' };
  const reviewMessages: AgenticChatMessage[] = [
    ...currentMessages,
    {
      role: 'system',
      content: `Review this heuristic tool activation.
Tool: ${candidate.toolName}
User request: ${userMessage}
Suggested args: ${JSON.stringify(candidate.suggestedArgs)}
Thoth gate action hint: ${
        thothGateTelemetry
          ? thothGateTelemetryActionHint(thothGateTelemetry)
          : 'none'
      }

Respond with exactly one:
ACCEPT
MODIFY {"key":"value"}
CANCEL`,
    },
  ];
  const result = await runtime.infer({
    model: request.model,
    messages: reviewMessages,
    temperature: 0,
    maxTokens: 80,
    responseFormat: { type: 'text' },
    round: 0,
    purpose: 'tool-review',
  });
  const normalized = result.content.trim().toUpperCase();
  if (normalized.startsWith('CANCEL')) return { action: 'cancel' };
  if (normalized.startsWith('MODIFY')) {
    const json = result.content.match(/\{[\s\S]*\}/u);
    if (json) {
      try {
        return {
          action: 'modify',
          modifiedArgs: JSON.parse(json[0]) as Record<string, unknown>,
        };
      } catch {
        return { action: 'accept' };
      }
    }
  }
  return { action: 'accept' };
}

async function resolveToolArgs(
  request: AgenticChatRequest,
  runtime: AgenticChatRuntime,
  candidate: ToolCandidate,
  userMessage: string,
  currentMessages: AgenticChatMessage[],
): Promise<Record<string, unknown>> {
  if (Object.keys(candidate.suggestedArgs).length > 0) {
    return candidate.suggestedArgs;
  }
  const result = await runtime.infer({
    model: request.model,
    messages: [
      ...currentMessages,
      {
        role: 'system',
        content: `Craft JSON arguments for tool "${candidate.toolName}".
User request: ${userMessage}
Return only a JSON object.`,
      },
    ],
    temperature: 0,
    maxTokens: 120,
    responseFormat: { type: 'json_object' },
    round: 0,
    purpose: 'tool-args',
  });
  const json = result.content.match(/\{[\s\S]*\}/u);
  if (!json) return {};
  try {
    return JSON.parse(json[0]) as Record<string, unknown>;
  } catch {
    return {};
  }
}

async function executeToolWithValidation(args: {
  toolCall: AgenticToolCall;
  mcp?: AgenticMcpClient;
  runtime: AgenticChatRuntime;
  request: AgenticChatRequest;
  currentMessages: AgenticChatMessage[];
  originalUserMessage: string;
  validateToolResults: boolean;
  maxValidationRetries: number;
  trace: AgenticTraceEntry[];
}): Promise<ToolExecutionResult> {
  const parsedArgs = parseToolArguments(args.toolCall.function.arguments);
  const mcp = args.mcp ?? args.runtime.mcp;
  if (!mcp) {
    return {
      toolCallId: args.toolCall.id,
      toolName: args.toolCall.function.name,
      args: parsedArgs,
      result: null,
      success: false,
      durationMs: 0,
      error: 'No MCP client configured',
    };
  }

  let currentArgs = parsedArgs;
  let bestResult: ToolExecutionResult | null = null;
  let bestQuality = -1;

  for (let attempt = 0; attempt < args.maxValidationRetries; attempt += 1) {
    const result = await mcp.callTool(
      args.toolCall.function.name,
      currentArgs,
      { toolCallId: args.toolCall.id },
    );
    recordThothOathTelemetryTrace(args.trace, result);
    const validation = args.validateToolResults
      ? validateToolResult(result.toolName, currentArgs, result)
      : { valid: true, quality: 1, issues: [], suggestRetry: false };
    args.trace.push({
      layer: 'C3',
      message: 'Tool result validated',
      data: {
        toolName: result.toolName,
        quality: roundScore(validation.quality),
        valid: validation.valid,
        issues: validation.issues,
        attempt: attempt + 1,
      },
    });

    if (validation.quality > bestQuality) {
      bestQuality = validation.quality;
      bestResult = result;
    }
    if (validation.valid || !validation.suggestRetry) {
      return result;
    }
    if (validation.modifiedArgs) {
      currentArgs = validation.modifiedArgs;
      continue;
    }
    if (attempt < args.maxValidationRetries - 1) {
      currentArgs = await refineToolArgsAfterValidation(
        args.request,
        args.runtime,
        args.currentMessages,
        args.originalUserMessage,
        result.toolName,
        currentArgs,
        validation,
      );
    }
  }

  return bestResult!;
}

function recordThothOathTelemetryTrace(
  trace: AgenticTraceEntry[],
  result: ToolExecutionResult,
): void {
  const payload = isRecord(result.result) ? result.result : null;
  const rawEvents = payload?.telemetryEvents ?? payload?.telemetryEvent;
  const events = Array.isArray(rawEvents) ? rawEvents : rawEvents ? [rawEvents] : [];
  for (const event of events.filter(isRecord)) {
    const kind = typeof event.kind === 'string' ? event.kind : 'unknown';
    trace.push({
      layer: 'C4',
      message: `Thoth oath telemetry observed: ${kind}`,
      data: {
        toolName: result.toolName,
        kind,
        oathId: typeof event.oathId === 'string' ? event.oathId : null,
        effectiveStatus:
          typeof event.effectiveStatus === 'string' ? event.effectiveStatus : null,
        activeOathCount:
          typeof event.activeOathCount === 'number' ? event.activeOathCount : null,
        nonAuthoritative: event.nonAuthoritative === true,
        theoremRef: typeof event.theoremRef === 'string' ? event.theoremRef : null,
      },
    });
  }
}

function validateToolResult(
  toolName: string,
  toolArgs: Record<string, unknown>,
  result: ToolExecutionResult,
): ResultValidation {
  if (!result.success || result.error) {
    return {
      valid: false,
      quality: 0,
      issues: [result.error ?? 'Tool execution failed'],
      suggestRetry: isRetryableToolError(result.error),
    };
  }
  if (result.result === null || result.result === undefined) {
    return {
      valid: false,
      quality: 0,
      issues: ['Tool returned no result'],
      suggestRetry: true,
    };
  }

  const serialized =
    typeof result.result === 'string'
      ? result.result
      : JSON.stringify(result.result);
  if (serialized.length < 3) {
    return {
      valid: false,
      quality: 0.2,
      issues: ['Tool result is too short to be useful'],
      suggestRetry: true,
    };
  }
  if (toolName.includes('search') && !hasSearchSignal(result.result)) {
    return {
      valid: false,
      quality: 0.45,
      issues: ['Search-like tool returned no recognizable hits'],
      suggestRetry: false,
    };
  }
  if (
    toolName.includes('summarize') &&
    typeof toolArgs.text === 'string' &&
    serialized.length >= toolArgs.text.length
  ) {
    return {
      valid: false,
      quality: 0.5,
      issues: ['Summary is not shorter than source text'],
      suggestRetry: false,
    };
  }

  return { valid: true, quality: 1, issues: [], suggestRetry: false };
}

function isRetryableToolError(error: string | undefined): boolean {
  if (!error) return false;
  const normalized = error.toLowerCase();
  return (
    normalized.includes('timeout') ||
    normalized.includes('rate') ||
    normalized.includes('invalid') ||
    normalized.includes('too large')
  );
}

function hasSearchSignal(value: unknown): boolean {
  if (Array.isArray(value)) return value.length > 0;
  if (!isRecord(value)) return true;
  for (const key of ['results', 'tools', 'backends', 'matches', 'items']) {
    const field = value[key];
    if (Array.isArray(field) && field.length > 0) return true;
  }
  return Object.keys(value).length > 0;
}

async function refineToolArgsAfterValidation(
  request: AgenticChatRequest,
  runtime: AgenticChatRuntime,
  currentMessages: AgenticChatMessage[],
  originalUserMessage: string,
  toolName: string,
  previousArgs: Record<string, unknown>,
  validation: ResultValidation,
): Promise<Record<string, unknown>> {
  const result = await runtime.infer({
    model: request.model,
    messages: [
      ...currentMessages,
      {
        role: 'system',
        content: `Tool "${toolName}" returned a low-quality result.
Issues: ${validation.issues.join(', ')}
Previous args: ${JSON.stringify(previousArgs)}
User request: ${originalUserMessage}

Return improved arguments as JSON only.`,
      },
    ],
    temperature: 0,
    maxTokens: 120,
    responseFormat: { type: 'json_object' },
    round: 0,
    purpose: 'tool-retry',
  });
  const json = result.content.match(/\{[\s\S]*\}/u);
  if (!json) return previousArgs;
  try {
    return JSON.parse(json[0]) as Record<string, unknown>;
  } catch {
    return previousArgs;
  }
}

function appendToolExchange(
  messages: AgenticChatMessage[],
  toolCall: AgenticToolCall,
  result: ToolExecutionResult,
): void {
  messages.push({
    role: 'assistant',
    content: null,
    tool_calls: [toolCall],
  });
  messages.push({
    role: 'tool',
    tool_call_id: toolCall.id,
    name: result.toolName,
    content: serializeToolResult(result),
  });
}

function serializeToolResult(result: ToolExecutionResult): string {
  if (result.error) {
    return JSON.stringify({ error: result.error });
  }
  return typeof result.result === 'string'
    ? result.result
    : JSON.stringify(result.result);
}

function parseToolArguments(value: string): Record<string, unknown> {
  try {
    const parsed = JSON.parse(value);
    return isRecord(parsed) ? parsed : {};
  } catch {
    return {};
  }
}

function buildToolCall(
  name: string,
  toolArgs: Record<string, unknown>,
  randomId: (prefix: string) => string,
): AgenticToolCall {
  return {
    id: randomId('call'),
    type: 'function',
    function: {
      name,
      arguments: JSON.stringify(toolArgs),
    },
  };
}

function parseToolCalls(
  output: string,
  validToolNames?: Set<string>,
): { text: string; toolCalls: AgenticToolCall[] | null } {
  const toolCalls: AgenticToolCall[] = [];
  let text = output;
  const blockRegex = /```(?:tool_call|json)?\s*\n?\s*(\{[\s\S]*?\})\s*\n?```/gu;
  let match: RegExpExecArray | null;
  while ((match = blockRegex.exec(output)) !== null) {
    const parsed = parseToolCallObject(match[1]);
    if (parsed && (!validToolNames || validToolNames.has(parsed.name))) {
      toolCalls.push({
        id: createId('call'),
        type: 'function',
        function: {
          name: parsed.name,
          arguments: JSON.stringify(parsed.arguments),
        },
      });
      text = text.replace(match[0], '').trim();
    }
  }

  if (toolCalls.length === 0) {
    const rawRegex =
      /\{\s*"name"\s*:\s*"([^"]+)"\s*,\s*"arguments"\s*:\s*(\{[^}]*\})\s*\}/gu;
    while ((match = rawRegex.exec(output)) !== null) {
      const name = match[1]!;
      if (validToolNames && !validToolNames.has(name)) continue;
      try {
        const args = JSON.parse(match[2]!);
        toolCalls.push({
          id: createId('call'),
          type: 'function',
          function: {
            name,
            arguments: JSON.stringify(isRecord(args) ? args : {}),
          },
        });
        text = text.replace(match[0], '').trim();
      } catch {
        // Ignore malformed raw tool JSON.
      }
    }
  }

  return {
    text: text.trim(),
    toolCalls: toolCalls.length > 0 ? toolCalls : null,
  };
}

function parseToolCallObject(
  value: string | undefined,
): { name: string; arguments: Record<string, unknown> } | null {
  if (!value) return null;
  try {
    const parsed = JSON.parse(value);
    if (!isRecord(parsed) || typeof parsed.name !== 'string') return null;
    return {
      name: parsed.name,
      arguments: isRecord(parsed.arguments) ? parsed.arguments : {},
    };
  } catch {
    return null;
  }
}

async function evaluateIntentSatisfaction(
  request: AgenticChatRequest,
  runtime: AgenticChatRuntime,
  originalUserMessage: string,
  currentMessages: AgenticChatMessage[],
  lastToolResults: ToolExecutionResult[],
): Promise<IntentEvaluation> {
  const toolSummary =
    lastToolResults.length === 0
      ? '(no tool results)'
      : lastToolResults
          .map((result) =>
            `${result.toolName}: ${
              result.error ? `ERROR ${result.error}` : serializeToolResult(result)
            }`,
          )
          .join('\n');
  const result = await runtime.infer({
    model: request.model,
    messages: [
      {
        role: 'system',
        content:
          'Evaluate whether the assistant has fully satisfied the user request. Respond with JSON only.',
      },
      {
        role: 'user',
        content: `Original request:
${originalUserMessage}

Conversation:
${renderMessagesForPrompt(currentMessages).slice(-6000)}

Tool results:
${toolSummary}

Return {"satisfied": boolean, "confidence": number, "reasoning": string, "suggestedNextAction": string}.`,
      },
    ],
    temperature: 0,
    maxTokens: 160,
    responseFormat: { type: 'json_object' },
    round: 0,
    purpose: 'intent-evaluation',
  });
  const json = result.content.match(/\{[\s\S]*\}/u);
  if (!json) {
    return inferIntentFromText(result.content);
  }
  try {
    const parsed = JSON.parse(json[0]);
    return {
      satisfied: Boolean(parsed.satisfied),
      confidence:
        typeof parsed.confidence === 'number' ? parsed.confidence : 0.5,
      reasoning:
        typeof parsed.reasoning === 'string'
          ? parsed.reasoning
          : 'No reasoning provided',
      ...(typeof parsed.suggestedNextAction === 'string'
        ? { suggestedNextAction: parsed.suggestedNextAction }
        : {}),
    };
  } catch {
    return inferIntentFromText(result.content);
  }
}

function inferIntentFromText(value: string): IntentEvaluation {
  const normalized = value.toLowerCase();
  if (
    normalized.includes('satisfied') ||
    normalized.includes('complete') ||
    normalized.includes('fully addressed')
  ) {
    return { satisfied: true, confidence: 0.7, reasoning: value.slice(0, 120) };
  }
  return {
    satisfied: false,
    confidence: 0.5,
    reasoning: value.slice(0, 120) || 'Unable to determine satisfaction',
  };
}

function continuationNudgeWithTelemetry(
  evaluation: IntentEvaluation,
  metrics: ThothGateTelemetryMetrics | undefined,
): string {
  const base = continuationNudge(evaluation);
  if (!metrics) return base;
  return `${base}
Thoth gate hint: ${thothGateTelemetryActionHint(metrics)}. Keep any action non-authoritative and source-backed.`;
}

function continuationNudge(evaluation: IntentEvaluation): string {
  if (evaluation.suggestedNextAction) {
    return `The user's request may not be fully addressed yet. Consider: ${evaluation.suggestedNextAction}`;
  }
  return "The user's request may not be fully addressed yet. Continue with the next useful step.";
}

function getLastUserMessage(messages: AgenticChatMessage[]): string {
  for (let index = messages.length - 1; index >= 0; index -= 1) {
    const message = messages[index]!;
    if (message.role === 'user' && typeof message.content === 'string') {
      return message.content;
    }
  }
  return '';
}

function estimateTokensFromMessages(messages: AgenticChatMessage[]): number {
  return estimateTokens(renderMessagesForPrompt(messages));
}

function estimateTokens(value: string): number {
  return Math.max(1, Math.ceil(value.length / 4));
}

function roundScore(value: number): number {
  return Math.round(value * 1000) / 1000;
}

function createId(prefix: string): string {
  if (typeof crypto !== 'undefined' && typeof crypto.randomUUID === 'function') {
    return `${prefix}_${crypto.randomUUID().slice(0, 12)}`;
  }
  return `${prefix}_${Date.now().toString(36)}_${Math.random()
    .toString(36)
    .slice(2, 8)}`;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}
