// @a0n/gnosis/distributed-inference-host/src/agentic-economy.ts
/**
 * The Aeon-12 tool economy: a visible stamina/score/syzygy layer for this
 * package's agentic tool-calling loop (`agentic-chat.ts`), ported from
 * `open-source/gnosis/moonshine/src/agent_economy.rs` (Rust). There is no
 * shared crate/package binding the two implementations together — keep the
 * numeric tables in sync by hand if either changes.
 *
 * The gauge itself is `godWeight(R, v) = R - min(v, R) + 1`, Lean-proved in
 * `open-source/gnosis-math/Gnosis/GodFormula.lean` (conservation, ceiling,
 * floor, antitonicity) and specialized at `R = 12` ("Aeon-12") the way
 * `Gnosis.AeonTwelveCarrierList.GodTwelveSlice` does. `godWeight`'s name and
 * signature follow the existing TS ports already in this monorepo
 * (`open-source/aeon-3d/src/entropy-engine.ts:204`,
 * `open-source/gnosis/src/organism-repair.ts:260`) rather than inventing a
 * third convention. Citation-token overlap for the finisher bonus reuses
 * `agentic-chat.ts`'s own `tokenizeToolText` rather than a second tokenizer.
 *
 * `agentic-chat.ts`'s own tool-candidate scoring (`scoreDerivedToolMatch`,
 * `selectVisibleTools`) does not share the two structural bugs Moonshine's
 * second refinement pass fixed (a raw-JSON-substring scoring tier, and a
 * positional eviction step) — it only scores against a tool's name/description
 * and fills candidates additively with no eviction — so only the *new*
 * capability (this economy layer) needed porting, not those bug fixes.
 */

import { tokenizeToolText } from './agentic-chat';

export const AEON12_R = 12;

/**
 * Lean: god_formula — w = R - min(v, R) + 1
 * The +1 is the clinamen (sliver). Zero is impossible.
 */
export function godWeight(rounds: number, rejections: number): number {
  return rounds - Math.min(rejections, rounds) + 1;
}

export type RiskTier = 0 | 1 | 2;

const DELETE_VERBS = ['delete', 'remove', 'wipe', 'drop', 'destroy', 'unlink', 'purge'];
const WRITE_VERBS = [
  'write',
  'replace',
  'patch',
  'edit',
  'create',
  'move',
  'rename',
  'append',
  'mkdir',
];
const VCS_VERBS = ['commit', 'push', 'checkout', 'reset', 'rebase'];

/**
 * Namespace/verb fallback risk classifier — the simplest viable port of
 * Moonshine's `classify_tool_risk` (`agent_tools.rs`). `AgenticToolDefinition`
 * carries no `x-forkjoin` stage metadata today, so only the name-convention
 * tier ports; tier 0 = read_only, 1 = workspace_write/formal_claim,
 * 2 = destructive/vcs_mutation/network.
 */
export function classifyToolRisk(toolName: string): RiskTier {
  const name = toolName.toLowerCase();
  const isVcs = name.includes('git') && VCS_VERBS.some((verb) => name.includes(verb));
  if (isVcs) return 2;
  if (DELETE_VERBS.some((verb) => name.includes(verb))) return 2;
  if (WRITE_VERBS.some((verb) => name.includes(verb))) return 1;
  if (name.includes('thm_') || name.includes('lean') || name.includes('proof')) return 1;
  return 0;
}

/** Vent cost by tier and outcome. Mirrors `agent_economy.rs`'s `vent_cost` table. */
function ventCost(tier: RiskTier, success: boolean): number {
  const table: Record<RiskTier, { success: number; failure: number }> = {
    0: { success: 0, failure: 1 },
    1: { success: 1, failure: 3 },
    2: { success: 2, failure: 6 },
  };
  return success ? table[tier].success : table[tier].failure;
}

/** Base score per successful call, before the syzygy multiplier. */
function tierScore(tier: RiskTier): number {
  return { 0: 5, 1: 10, 2: 15 }[tier];
}

/** Syzygy multiplier in tenths (10 = x1.0 .. 20 = x2.0, capped). */
function syzygyMultiplierTenths(chainLen: number): number {
  if (chainLen <= 1) return 10;
  if (chainLen === 2) return 15;
  return 20;
}

/** Deterministic FNV-1a hash, used only to place a tool's domain on the Aeon-12 dial. */
function fnv1a(input: string): number {
  let hash = 0x811c9dc5;
  for (let i = 0; i < input.length; i += 1) {
    hash ^= input.charCodeAt(i);
    hash = Math.imul(hash, 0x01000193);
  }
  return hash >>> 0;
}

/**
 * Place a tool on the Fin-12 dial by its namespace prefix (the part of its
 * name before the first `/` or `_`). `AgenticToolDefinition` has no
 * `dominantDomain`/`domains` field the way Moonshine's `ToolMeta` does, so
 * this falls back to the tool's own namespace/name prefix.
 */
function domainBucket(toolName: string): number {
  const namespace = toolName.split(/[/_]/u)[0]?.toLowerCase() ?? toolName.toLowerCase();
  return fnv1a(namespace) % 12;
}

/** Circular distance on Fin-12, the same shape as `Gnosis.AeonCycleTwelveShadow.shortChord`. */
function shortChordDistance(a: number, b: number): number {
  const d = Math.abs(a - b);
  return Math.min(d, 12 - d);
}

type Transition = 'first' | 'strong' | 'weak' | 'break';

function classifyTransition(
  last: { bucket: number; tier: RiskTier } | null,
  bucket: number,
  tier: RiskTier,
): Transition {
  if (!last) return 'first';
  const distance = shortChordDistance(last.bucket, bucket);
  if (distance <= 1 && tier >= last.tier) return 'strong';
  if (distance <= 3) return 'weak';
  return 'break';
}

/** Cue-token set for citation comparison — "citing" a tool result means sharing meaningful tokens, not stopwords. */
function citationTokens(text: string): Set<string> {
  return new Set(tokenizeToolText(text));
}

function intersectionSize(a: Set<string>, b: Set<string>): number {
  let count = 0;
  for (const token of a) {
    if (b.has(token)) count += 1;
  }
  return count;
}

/**
 * Canonical `(tool name, arguments)` signature for repeat-call detection.
 * `JSON.stringify` on a plain object does not guarantee key order across
 * differently-constructed-but-equal objects — sort keys first so an
 * identical logical call always produces the same signature string.
 */
function callSignature(toolName: string, args: Record<string, unknown>): string {
  const sortedKeys = Object.keys(args).sort();
  const canonical: Record<string, unknown> = {};
  for (const key of sortedKeys) canonical[key] = args[key];
  return `${toolName}::${JSON.stringify(canonical)}`;
}

/** Flat bonus for finishing on a cited result. Never scaled by the syzygy multiplier or chain length. */
export const FINISHER_BONUS = 12;

export interface EconomySummary {
  weight: number;
  ceiling: number;
  score: number;
  winded: boolean;
  lastEvent: string;
}

/**
 * One request's running Aeon-12 tool-economy state. `vent` only ever
 * increases within a request — `godWeight` is proved antitone, so no regen
 * step is needed to stay formally clean; the gauge narrows what can be
 * called as it drains but (per `godWeight`'s floor) never truly reaches zero.
 */
export class AgentEconomy {
  private readonly maxStamina: number;
  private vent = 0;
  private chainLen = 0;
  private last: { bucket: number; tier: RiskTier } | null = null;
  private lastResultTokens: Set<string> | null = null;
  private lastCall: string | null = null;

  score = 0;
  winded = false;
  lastEvent = '';

  constructor(maxStamina: number = AEON12_R) {
    this.maxStamina = Math.max(1, maxStamina);
  }

  weight(): number {
    return godWeight(this.maxStamina, Math.min(this.vent, this.maxStamina));
  }

  ceiling(): number {
    return this.maxStamina + 1;
  }

  /**
   * True before any tool call has succeeded or failed this request (the
   * baseline: full power, no score, no chain). At rest the status block
   * would only ever read the same static boilerplate, so callers skip
   * injecting it — one less thing for a small/weak model to latch onto and
   * echo back instead of answering or calling a tool.
   */
  isAtRest(): boolean {
    return this.vent === 0 && this.score === 0;
  }

  summary(): EconomySummary {
    return {
      weight: this.weight(),
      ceiling: this.ceiling(),
      score: this.score,
      winded: this.winded,
      lastEvent: this.lastEvent,
    };
  }

  private applyVent(amount: number): void {
    this.vent = Math.min(this.vent + amount, this.maxStamina * 2);
    this.winded = this.vent >= this.maxStamina;
  }

  recordSuccess(
    toolName: string,
    risk: RiskTier,
    args: Record<string, unknown>,
    resultText: string,
  ): void {
    this.applyVent(ventCost(risk, true));

    const signature = callSignature(toolName, args);
    const isRepeat = this.lastCall === signature;
    this.lastCall = signature;
    if (isRepeat) {
      // The tool call really did succeed (ordinary success-tier vent already
      // applied above), but an identical repeat produced no NEW information —
      // score it like a no-op and break the chain so it can't be laundered
      // into a "productive" transition.
      this.chainLen = 0;
      this.last = null;
      this.lastResultTokens = citationTokens(resultText);
      this.lastEvent = '+0 (identical repeat call — no new information, chain broken)';
      return;
    }

    const bucket = domainBucket(toolName);
    const transition = classifyTransition(this.last, bucket, risk);
    this.chainLen = transition === 'break' || transition === 'first' ? 1 : this.chainLen + 1;
    const multiplierTenths = transition === 'strong' ? syzygyMultiplierTenths(this.chainLen) : 10;
    const gained = Math.floor((tierScore(risk) * multiplierTenths) / 10);
    this.score += gained;
    this.last = { bucket, tier: risk };
    this.lastResultTokens = citationTokens(resultText);
    this.lastEvent = `+${gained} (syzygy x${(multiplierTenths / 10).toFixed(1)}, chain ${this.chainLen})`;
  }

  recordFailure(risk: RiskTier): void {
    this.applyVent(ventCost(risk, false));
    this.chainLen = 0;
    this.last = null;
    this.lastCall = null;
    this.lastEvent = '+0 (failed call — no gain, chain broken)';
  }

  /**
   * Awarded once, when the final plain-text answer actually cites the most
   * recent successful tool result. `recordSuccess` only ever rewards calling
   * MORE tools (via the syzygy chain) — this is the counterweight that
   * rewards stopping. Returns the amount gained (0 if no tool has succeeded
   * yet this request, or the answer doesn't cite the result).
   */
  awardFinisherIfCiting(answer: string): number {
    if (!this.lastResultTokens || this.lastResultTokens.size === 0) return 0;
    const answerTokens = citationTokens(answer);
    const overlap = intersectionSize(answerTokens, this.lastResultTokens);
    // Adaptive floor: a thin result (1 cue token) can't demand 2 shared
    // tokens, but a richer result needs >= 2 so one incidental shared word
    // doesn't trivially satisfy it.
    if (overlap < Math.min(this.lastResultTokens.size, 2)) return 0;
    this.score += FINISHER_BONUS;
    this.lastEvent = `+${FINISHER_BONUS} (finisher — cited the tool result, ${overlap} shared token(s))`;
    return FINISHER_BONUS;
  }
}

/**
 * The game-framed block injected into the prompt once a round has real state
 * to report (`!economy.isAtRest()`). Carrot framing is deliberate: there is
 * no forced tool-choice enforcement available, so this is the only lever
 * that can make the model prefer calling a tool over narrating one.
 */
export function economyStatusBlock(economy: AgentEconomy): string {
  const weight = economy.weight();
  const ceiling = economy.ceiling();
  const lines = [
    'AEON-12 GAME STATE (real — it changes what you can do):',
    `Power: ${weight}/${ceiling} | Score: ${economy.score} pts`,
    '- Every tool call vents the dial. A FAILED call vents far more than a success, ' +
      'and mutating/destructive/network calls vent the most when they fail — read ' +
      "before you write. The dial can never hit true zero, but it narrows what you're " +
      'allowed to touch as it drains.',
    '- Calling tools in the same neighborhood back to back (discover -> inspect -> ' +
      'mutate) is a SYZYGY: nearby chains multiply your score up to x2. A wide jump to ' +
      'an unrelated tool breaks the chain back to x1.',
    '- FINISHING pays too: once a tool call has succeeded, a final answer that cites ' +
      `its specific result earns a flat +${FINISHER_BONUS} finisher bonus. You don't need ` +
      'another tool call just to keep a chain going — cite the result and answer.',
  ];
  if (economy.lastEvent) {
    lines.push(`- Last action: ${economy.lastEvent}`);
  }
  if (economy.winded) {
    lines.push(
      `\nTHE DIAL IS SPENT (power ${weight}/${ceiling}). Only read-only tools are ` +
        'offered until you finish this answer — you cannot mutate anything right now.',
    );
  }
  return lines.join('\n');
}
