// @a0n/gnosis/distributed-inference-host/src/amplituhedron-wire.ts
/**
 * amplituhedron-wire.ts вҖ” wire codec for the Amplituhedron volume-cache
 * coordinator protocol (Death #2 prefill-skip).
 *
 * SCOPE
 * -----
 * Encodes / decodes the binary envelope exchanged between the host
 * coordinator (`pipeline.ts`) and the per-station `/amplituhedron/replay`
 * and `/amplituhedron/capture` endpoints. The envelope wraps a tail-residual
 * f32 stream plus the prefix-hash / layer-range metadata the kernel needs
 * to key its capture/replay table.
 *
 * REFERENCES
 * ----------
 * - `AMPLITUHEDRON_COORDINATOR_PROTOCOL.md` В§4 вҖ” coordinator flow (compute
 *   prefix hash, fan out replay, branch on all-hit vs any-miss, fan out
 *   capture on miss).
 * - `open-source/gnosis/distributed-inference/AMPLITUHEDRON_PLAN.md` В§3 вҖ”
 *   the cache-key shape (`prefix_hash` + `prefix_len` + `layer_lo..layer_hi`).
 * - `apps/distributed-inference-worker/src/index.ts` вҖ” station-side handler
 *   for these endpoints (sibling Agent A spec).
 *
 * WIRE FORMAT
 * -----------
 *   в”Ңв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”¬в”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”¬в”Җв”Җв”Җв”Җв”Җв”Җв”Җв”¬в”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”җ
 *   в”Ӯ magic  в”Ӯ version в”Ӯ flags в”Ӯ headers + payload                    в”Ӯ
 *   в”Ӯ 'BWAH' в”Ӯ 0x01    в”Ӯ u8    в”Ӯ (replay:   tail_len + kv_slab_len)   в”Ӯ
 *   в”Ӯ 4 B    в”Ӯ 1 B     в”Ӯ 1 B   в”Ӯ (capture:  + prefix_hash + prefix_   в”Ӯ
 *   в”Ӯ        в”Ӯ         в”Ӯ       в”Ӯ            len + layer_lo + layer_hi)в”Ӯ
 *   в””в”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”ҙв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”ҙв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”ҙв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”ҳ
 *
 *   flags: bit 0 = hit, bit 1 = tail_residual present,
 *          bit 2 = prefix_residual present, bits 3..7 reserved.
 *   All multi-byte integers are little-endian.
 *
 * PAYLOAD CODEC
 * -------------
 * The payload section is a `.bw` framed blob produced by
 * `@a0n/bitwise/bw-codec::encodeBw(rawLeBytes, "application/octet-stream")`.
 * This is the canonical bitwise wire вҖ” same format-compliance and fp48
 * integrity carrier used everywhere else in the `.bw` ecosystem.
 *
 * The 2026-05-10 swap replaced the interim gzip layer (1.07-1.08Г— on
 * realistic workloads вҖ” random residuals, post-RMSNorm Luminary) with
 * the bw_codec wire. Wire size is comparable to raw LE on random data
 * (~14 bytes of framing overhead) but the wire is now uniform across
 * HTTP body bytes, captures, replays, and carries an fp48 integrity
 * fingerprint for free.
 *
 * BIT-EXACTNESS
 * -------------
 * f32 round-trip is bitwise-identical; the inner `bw_codec` round-trip
 * is byte-exact (`load_time_reconstruction_exact`), and the f32 вҶ” bytes
 * conversion is raw little-endian (no normalization, no quantization).
 * For a 3584-element Float32Array the raw payload is 14336 B; bw_codec
 * wraps it to ~14356 B (15 B header + manifest + packed index segment);
 * BWAH envelope adds 14 B вҶ’ total ~14370 B on the wire.
 *
 * CROSS-LANGUAGE
 * --------------
 * The Rust port `bitwise::bw_codec::encode_bw` / `decode_bw` mirrors this
 * TS source byte-for-byte. Worker (CF) and host (Node/Bun) produce
 * identical bytes on the wire by design.
 */

import { readFileSync } from "node:fs";

import { decodeBw, encodeBw } from "@a0n/bitwise/bw-codec";

export const AMPLITUHEDRON_CONTENT_TYPE = "application/x-bitwise-residual";

/** MIME type fed into the inner `bw_codec` blob. */
const PAYLOAD_BW_MIME = "application/octet-stream";

/**
 * Residual positional-dictionary loader.
 *
 * When `GNOSIS_RESIDUAL_DICT` points at a `.bwdict` file with the canonical
 * `BWDC` header, the encoder routes every residual through the bw_codec
 * positional-dictionary path (`BW_VERSION_POSDICT_FLAG`). When the env var
 * is unset or the file is malformed, the encoder falls back to plain bw
 * (version byte = 0x01).
 *
 * Layout (little-endian):
 *   в”Ңв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”¬в”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”¬в”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”¬в”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”¬в”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”җ
 *   в”Ӯ magic    в”Ӯ version в”Ӯ count   в”Ӯ entry_len в”Ӯ count Г— entry_len bytes в”Ӯ
 *   в”Ӯ 'BWDC'   в”Ӯ u32 = 1 в”Ӯ u32     в”Ӯ u32       в”Ӯ                         в”Ӯ
 *   в””в”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”ҙв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”ҙв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”ҙв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”ҙв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”Җв”ҳ
 *
 * The loaded entries are cached at module scope. Tests can reset the cache
 * via {@link _resetDictCache} so each test sees a fresh `GNOSIS_RESIDUAL_DICT`
 * value.
 *
 * In non-Node runtimes (`readFileSync` unavailable / throws) the loader
 * falls back to "no dictionary" silently вҖ” there is nothing it can do.
 */
const BWDC_MAGIC = new Uint8Array([0x42, 0x57, 0x44, 0x43]);
const BWDC_HEADER_LEN = 16;
const BWDC_VERSION = 1;

let __dictCache: Uint8Array[] | null | undefined;

/**
 * Reset the module-level dictionary cache. Intended for test fixtures
 * (e.g. `cross-lang-dict-fixture.ts`) that need to flip
 * `GNOSIS_RESIDUAL_DICT` between calls.
 */
export function _resetDictCache(): void {
  __dictCache = undefined;
}

function loadResidualDict(): Uint8Array[] | undefined {
  if (__dictCache !== undefined) {
    return __dictCache === null ? undefined : __dictCache;
  }
  let path: string | undefined;
  try {
    // `process` may be undefined in non-Node runtimes (jsdom, CF Worker).
    path = typeof process !== "undefined" ? process.env?.["GNOSIS_RESIDUAL_DICT"] : undefined;
  } catch {
    path = undefined;
  }
  if (!path) {
    __dictCache = null;
    return undefined;
  }
  try {
    const buf = readFileSync(path);
    if (buf.length < BWDC_HEADER_LEN) {
      __dictCache = null;
      return undefined;
    }
    for (let i = 0; i < BWDC_MAGIC.length; i++) {
      if (buf[i] !== BWDC_MAGIC[i]) {
        __dictCache = null;
        return undefined;
      }
    }
    const dv = new DataView(buf.buffer, buf.byteOffset, buf.byteLength);
    const version = dv.getUint32(4, true);
    if (version !== BWDC_VERSION) {
      __dictCache = null;
      return undefined;
    }
    const count = dv.getUint32(8, true);
    const entryLen = dv.getUint32(12, true);
    if (buf.length < BWDC_HEADER_LEN + count * entryLen) {
      __dictCache = null;
      return undefined;
    }
    const entries: Uint8Array[] = [];
    for (let i = 0; i < count; i++) {
      const start = BWDC_HEADER_LEN + i * entryLen;
      // Copy out of the readFileSync Buffer so subsequent fs traffic can't
      // mutate the cache.
      entries.push(new Uint8Array(buf.subarray(start, start + entryLen)));
    }
    __dictCache = entries;
    return entries;
  } catch {
    __dictCache = null;
    return undefined;
  }
}

function bwEncodeOptions(): { positionalDictionary: Uint8Array[] } | undefined {
  const dict = loadResidualDict();
  return dict ? { positionalDictionary: dict } : undefined;
}

function bwDecodeOptions(): { positionalDictionary: Uint8Array[] } | undefined {
  const dict = loadResidualDict();
  return dict ? { positionalDictionary: dict } : undefined;
}

/** Magic 'BWAH' вҖ” Bitwise Amplituhedron. */
const MAGIC: readonly [number, number, number, number] = [0x42, 0x57, 0x41, 0x48];
const MAGIC_LEN = 4;
const VERSION = 0x01;

/** Flag bits. */
const FLAG_HIT = 0x01;
const FLAG_TAIL_PRESENT = 0x02;
const FLAG_PREFIX_PRESENT = 0x04;
const FLAG_RESERVED_MASK = 0xf8;

/** Byte offsets / sizes for the replay envelope. */
const REPLAY_HEADER_LEN = MAGIC_LEN + 1 /* version */ + 1 /* flags */ + 4 /* tail_len */ + 4; /* kv_slab_len */

/** Byte offsets / sizes for the capture envelope (adds prefix_hash + prefix_len + layer_lo + layer_hi). */
const CAPTURE_EXTRA_LEN = 8 /* prefix_hash u64 */ + 4 /* prefix_len u32 */ + 2 /* layer_lo u16 */ + 2; /* layer_hi u16 */
const CAPTURE_HEADER_LEN = REPLAY_HEADER_LEN + CAPTURE_EXTRA_LEN;

export interface AmplituhedronReplayBinary {
  hit: boolean;
  tailLen: number;
  kvSlabLen: number;
  tailResidual?: Float32Array;
  tailFingerprint?: string;
  prefixResidualLen?: number;
  prefixResidual?: Float32Array;
  prefixFingerprint?: string;
}

/** Thrown when a buffer cannot be parsed as an Amplituhedron envelope. */
export class AmplituhedronWireError extends Error {
  constructor(message: string) {
    super(`[amplituhedron-wire] ${message}`);
    this.name = "AmplituhedronWireError";
  }
}

function writeMagicVersion(dv: DataView, offset: number): number {
  let p = offset;
  for (let i = 0; i < MAGIC_LEN; i++) {
    dv.setUint8(p++, MAGIC[i]!);
  }
  dv.setUint8(p++, VERSION);
  return p;
}

function readMagicVersion(dv: DataView): number {
  if (dv.byteLength < MAGIC_LEN + 1) {
    throw new AmplituhedronWireError(
      `buffer too short for magic+version: ${dv.byteLength} < ${MAGIC_LEN + 1}`,
    );
  }
  for (let i = 0; i < MAGIC_LEN; i++) {
    if (dv.getUint8(i) !== MAGIC[i]!) {
      throw new AmplituhedronWireError(
        `wrong magic: expected 'BWAH' (0x${MAGIC.map((b) => b.toString(16).padStart(2, "0")).join("")}), got 0x${
          [0, 1, 2, 3].map((i) => dv.getUint8(i).toString(16).padStart(2, "0")).join("")
        }`,
      );
    }
  }
  const version = dv.getUint8(MAGIC_LEN);
  if (version !== VERSION) {
    throw new AmplituhedronWireError(
      `unsupported version: expected 0x${VERSION.toString(16).padStart(2, "0")}, got 0x${version.toString(16).padStart(2, "0")}`,
    );
  }
  return MAGIC_LEN + 1;
}

/**
 * Serialize the f32 samples as raw little-endian bytes, then wrap with
 * the canonical `bw_codec` wire.
 *
 * Routes through the positional-dictionary path when
 * `GNOSIS_RESIDUAL_DICT` is configured; otherwise plain bw.
 */
function f32ToBwPayload(samples: Float32Array): Uint8Array {
  const raw = new Uint8Array(samples.length * 4);
  const dv = new DataView(raw.buffer, raw.byteOffset, raw.byteLength);
  for (let i = 0; i < samples.length; i++) {
    dv.setFloat32(i * 4, samples[i]!, true /* little-endian */);
  }
  const opts = bwEncodeOptions();
  return opts ? encodeBw(raw, PAYLOAD_BW_MIME, opts) : encodeBw(raw, PAYLOAD_BW_MIME);
}

function bwPayloadToF32(payload: Uint8Array, length: number): Float32Array {
  let raw: Uint8Array;
  try {
    const opts = bwDecodeOptions();
    const decoded = opts ? decodeBw(payload, opts) : decodeBw(payload);
    raw = decoded.bytes;
  } catch (err) {
    throw new AmplituhedronWireError(
      `bw_codec decode failed: ${(err as Error).message}`,
    );
  }
  if (raw.byteLength !== length * 4) {
    throw new AmplituhedronWireError(
      `payload byte mismatch: decoded ${raw.byteLength} bytes, expected ${length * 4} (= 4 * ${length})`,
    );
  }
  const dv = new DataView(raw.buffer, raw.byteOffset, raw.byteLength);
  const out = new Float32Array(length);
  for (let i = 0; i < length; i++) {
    out[i] = dv.getFloat32(i * 4, true /* little-endian */);
  }
  return out;
}

/**
 * Handles the gnosis fingerprint Float32 Array workflow.
 */
export function fingerprintFloat32Array(samples: Float32Array): string {
  let h = 0xcbf29ce484222325n;
  const bytes = new Uint8Array(samples.buffer, samples.byteOffset, samples.byteLength);
  for (let i = 0; i < bytes.length; i++) {
    h ^= BigInt(bytes[i]!);
    h = (h * 0x100000001b3n) & 0xffffffffffffffffn;
  }
  return h.toString(16).padStart(16, '0');
}

/**
 * Encode an Amplituhedron replay response.
 *
 * When `hit` is false the body is header-only (no payload). When `hit` is
 * true `tailResidual` must be present and its length must equal `tailLen`.
 */
export function encodeReplayResponse(r: AmplituhedronReplayBinary): Uint8Array {
  if (r.hit) {
    if (!r.tailResidual) {
      throw new AmplituhedronWireError(
        "encodeReplayResponse: hit=true requires tailResidual",
      );
    }
    if (r.tailResidual.length !== r.tailLen) {
      throw new AmplituhedronWireError(
        `encodeReplayResponse: tailResidual.length=${r.tailResidual.length} != tailLen=${r.tailLen}`,
      );
    }
  }
  if (r.prefixResidual !== undefined) {
    if (!r.hit) {
      throw new AmplituhedronWireError(
        "encodeReplayResponse: prefixResidual requires hit=true",
      );
    }
    const prefixResidualLen = r.prefixResidualLen ?? r.prefixResidual.length;
    if (r.prefixResidual.length !== prefixResidualLen) {
      throw new AmplituhedronWireError(
        `encodeReplayResponse: prefixResidual.length=${r.prefixResidual.length} != prefixResidualLen=${prefixResidualLen}`,
      );
    }
  }

  let flags = 0;
  if (r.hit) flags |= FLAG_HIT;
  const tailPresent = r.hit && r.tailResidual !== undefined;
  const prefixPresent = r.hit && r.prefixResidual !== undefined;
  let tailPayloadRegion: Uint8Array = new Uint8Array(0);
  let prefixPayloadRegion: Uint8Array = new Uint8Array(0);
  if (tailPresent) {
    flags |= FLAG_TAIL_PRESENT;
    tailPayloadRegion = f32ToBwPayload(r.tailResidual!);
    r.tailFingerprint ??= fingerprintFloat32Array(r.tailResidual!);
  }
  if (prefixPresent) {
    flags |= FLAG_PREFIX_PRESENT;
    prefixPayloadRegion = f32ToBwPayload(r.prefixResidual!);
    r.prefixFingerprint ??= fingerprintFloat32Array(r.prefixResidual!);
  }

  const extensionLen = prefixPresent ? 8 : 0;
  const out = new Uint8Array(
    REPLAY_HEADER_LEN +
      extensionLen +
      tailPayloadRegion.byteLength +
      prefixPayloadRegion.byteLength,
  );
  const dv = new DataView(out.buffer, out.byteOffset, out.byteLength);

  let p = writeMagicVersion(dv, 0);
  dv.setUint8(p++, flags);
  dv.setUint32(p, r.tailLen >>> 0, true);
  p += 4;
  dv.setUint32(p, r.kvSlabLen >>> 0, true);
  p += 4;

  if (prefixPresent) {
    dv.setUint32(p, tailPayloadRegion.byteLength >>> 0, true);
    p += 4;
    dv.setUint32(p, (r.prefixResidualLen ?? r.prefixResidual!.length) >>> 0, true);
    p += 4;
  }
  if (tailPresent) {
    out.set(tailPayloadRegion, p);
    p += tailPayloadRegion.byteLength;
  }
  if (prefixPresent) {
    out.set(prefixPayloadRegion, p);
  }

  return out;
}

/**
 * Decode an Amplituhedron replay response.
 *
 * Throws `AmplituhedronWireError` on wrong magic, wrong version, reserved
 * flag bits set, or truncated payload.
 */
export function decodeReplayResponse(buf: Uint8Array): AmplituhedronReplayBinary {
  if (buf.length < REPLAY_HEADER_LEN) {
    throw new AmplituhedronWireError(
      `replay buffer too short: ${buf.length} < ${REPLAY_HEADER_LEN}`,
    );
  }
  const dv = new DataView(buf.buffer, buf.byteOffset, buf.byteLength);

  let p = readMagicVersion(dv);
  const flags = dv.getUint8(p++);
  if ((flags & FLAG_RESERVED_MASK) !== 0) {
    throw new AmplituhedronWireError(
      `reserved flag bits set: 0x${flags.toString(16).padStart(2, "0")}`,
    );
  }
  const tailLen = dv.getUint32(p, true);
  p += 4;
  const kvSlabLen = dv.getUint32(p, true);
  p += 4;

  const hit = (flags & FLAG_HIT) !== 0;
  const tailPresent = (flags & FLAG_TAIL_PRESENT) !== 0;
  const prefixPresent = (flags & FLAG_PREFIX_PRESENT) !== 0;
  let tailPayloadLen: number | undefined;
  let prefixResidualLen: number | undefined;

  if (prefixPresent) {
    if (!hit || !tailPresent) {
      throw new AmplituhedronWireError(
        "prefix residual replay payload requires hit and tail payload",
      );
    }
    if (buf.length < p + 8) {
      throw new AmplituhedronWireError("replay prefix extension header truncated");
    }
    tailPayloadLen = dv.getUint32(p, true);
    p += 4;
    prefixResidualLen = dv.getUint32(p, true);
    p += 4;
  }

  if (!tailPresent) {
    if (prefixPresent) {
      throw new AmplituhedronWireError("prefix residual without tail residual");
    }
    return { hit, tailLen, kvSlabLen };
  }

  // Payload region is the bw_codec-framed blob occupying everything after
  // the fixed header.
  const tailPayloadRegion =
    tailPayloadLen === undefined ? buf.subarray(p) : buf.subarray(p, p + tailPayloadLen);
  if (tailPayloadLen !== undefined && tailPayloadRegion.length !== tailPayloadLen) {
    throw new AmplituhedronWireError("replay tail payload truncated");
  }
  const tailResidual = bwPayloadToF32(tailPayloadRegion, tailLen);
  p += tailPayloadRegion.length;

  if (!prefixPresent) {
    return {
      hit,
      tailLen,
      kvSlabLen,
      tailResidual,
      tailFingerprint: fingerprintFloat32Array(tailResidual),
    };
  }

  const prefixPayloadRegion = buf.subarray(p);
  const prefixResidual = bwPayloadToF32(prefixPayloadRegion, prefixResidualLen!);
  return {
    hit,
    tailLen,
    kvSlabLen,
    tailResidual,
    tailFingerprint: fingerprintFloat32Array(tailResidual),
    prefixResidualLen,
    prefixResidual,
    prefixFingerprint: fingerprintFloat32Array(prefixResidual),
  };
}

/**
 * Encode an Amplituhedron capture request. The capture body always carries
 * a tail residual (FLAG_TAIL_PRESENT is always set; FLAG_HIT is unused for
 * captures and is kept zero so the wire layout stays in one shape).
 *
 * `prefixHash` is a decimal u64 string (matches the JSON / hex carriers used
 * in the wider mesh; BigInt parses any decimal u64 losslessly).
 */
export function encodeCaptureRequest(args: {
  prefixHash: string;
  prefixLen: number;
  layerLo: number;
  layerHi: number;
  tailResidual: Float32Array;
}): Uint8Array {
  const prefixHashBig = BigInt(args.prefixHash);
  if (prefixHashBig < 0n || prefixHashBig > 0xffffffffffffffffn) {
    throw new AmplituhedronWireError(
      `encodeCaptureRequest: prefixHash out of u64 range: ${args.prefixHash}`,
    );
  }
  if (args.layerLo < 0 || args.layerLo > 0xffff || args.layerHi < 0 || args.layerHi > 0xffff) {
    throw new AmplituhedronWireError(
      `encodeCaptureRequest: layer indices out of u16 range: lo=${args.layerLo} hi=${args.layerHi}`,
    );
  }

  const flags = FLAG_TAIL_PRESENT;
  const tailLen = args.tailResidual.length;
  const payloadRegion = f32ToBwPayload(args.tailResidual);
  const out = new Uint8Array(CAPTURE_HEADER_LEN + payloadRegion.byteLength);
  const dv = new DataView(out.buffer, out.byteOffset, out.byteLength);

  let p = writeMagicVersion(dv, 0);
  dv.setUint8(p++, flags);
  // Replay header section: tail_len (= residual length here) + kv_slab_len
  // (= 0 for captures; the kernel re-derives the slab from its own state).
  dv.setUint32(p, tailLen >>> 0, true);
  p += 4;
  dv.setUint32(p, 0, true);
  p += 4;

  // Capture extension: u64 prefix_hash, u32 prefix_len, u16 layer_lo, u16 layer_hi.
  dv.setBigUint64(p, prefixHashBig, true);
  p += 8;
  dv.setUint32(p, args.prefixLen >>> 0, true);
  p += 4;
  dv.setUint16(p, args.layerLo & 0xffff, true);
  p += 2;
  dv.setUint16(p, args.layerHi & 0xffff, true);
  p += 2;

  out.set(payloadRegion, p);
  return out;
}

/**
 * Handles the gnosis decode Capture Request workflow.
 */
export function decodeCaptureRequest(buf: Uint8Array): {
  prefixHash: string;
  prefixLen: number;
  layerLo: number;
  layerHi: number;
  tailResidual: Float32Array;
} {
  if (buf.length < CAPTURE_HEADER_LEN) {
    throw new AmplituhedronWireError(
      `capture buffer too short: ${buf.length} < ${CAPTURE_HEADER_LEN}`,
    );
  }
  const dv = new DataView(buf.buffer, buf.byteOffset, buf.byteLength);

  let p = readMagicVersion(dv);
  const flags = dv.getUint8(p++);
  if ((flags & FLAG_RESERVED_MASK) !== 0) {
    throw new AmplituhedronWireError(
      `reserved flag bits set: 0x${flags.toString(16).padStart(2, "0")}`,
    );
  }
  if ((flags & FLAG_TAIL_PRESENT) === 0) {
    throw new AmplituhedronWireError(
      "capture request must have FLAG_TAIL_PRESENT set",
    );
  }

  const tailLen = dv.getUint32(p, true);
  p += 4;
  // kv_slab_len ignored on capture вҖ” read past it for layout consistency.
  p += 4;

  const prefixHashBig = dv.getBigUint64(p, true);
  p += 8;
  const prefixLen = dv.getUint32(p, true);
  p += 4;
  const layerLo = dv.getUint16(p, true);
  p += 2;
  const layerHi = dv.getUint16(p, true);
  p += 2;

  const payloadRegion = buf.subarray(p);
  const tailResidual = bwPayloadToF32(payloadRegion, tailLen);

  return {
    prefixHash: prefixHashBig.toString(10),
    prefixLen,
    layerLo,
    layerHi,
    tailResidual,
  };
}
