// @a0n/gnosis/distributed-inference-host/src/amplituhedron-sharding.ts
/**
 * amplituhedron-sharding.ts — AlphaJump Multi-node Sharding.
 *
 * Implements the Maha-anchored sharding strategy for distributing
 * Amplituhedron volumes across a planetary-scale subnet.
 *
 * Anchors:
 *   - Maha (21): Sharding modulus. Term 8 of Taylor's Sequence.
 *   - Phyle Axis (3): Sub-shard grouping.
 *   - Axis Capacity (7): Terms per axis. Term 2 of Taylor's Sequence.
 *
 * Mirrors Gnosis.AlphaJumpTaylorSharding.lean.
 */

/** Maha (21): The Phyle Tripod stability anchor for sharding. */
export const MAHA_ANCHOR = 21;
const MAHA_ANCHOR_BIGINT = 21n;

/** Number of Phyle axes for balanced distribution. */
export const PHYLE_AXIS_COUNT = 3;

/** Capacity per axis shard. */
export const AXIS_SHARD_CAPACITY = 7;

export interface MahaShardRoute {
  shard: number;
  start: number;
  end: number;
}

export interface MahaPhysicalStationRange {
  stationIndex: number;
  shardStart: number;
  shardEnd: number;
}

/**
 * Maps a prefix hash to an AlphaJump shard index [0..20].
 *
 * Uses the BigInt modulus to ensure bit-exact alignment with the
 * Lean formalization and the Rust `u64 % 21` implementation.
 */
export function mapToMahaShard(prefixHash: string | bigint): number {
  const h = typeof prefixHash === 'string' ? BigInt(prefixHash) : prefixHash;
  // Maha (21) modulus ensures tripod stability.
  return Number(h % MAHA_ANCHOR_BIGINT);
}

/**
 * Maps a shard index to its primary Phyle Axis (0, 1, or 2).
 *
 * 21 = 3 * 7. Each axis carries 7 tripod terms.
 */
export function getShardAxis(shardIndex: number): number {
  return Math.floor(shardIndex / AXIS_SHARD_CAPACITY);
}

/**
 * Resolves the contiguous physical-station range assigned to a Maha shard.
 *
 * The cluster always maps onto 21 virtual Maha shards:
 * - 1 physical station owns all 21 virtual shards.
 * - 2 physical stations split them 11/10.
 * - 3 physical stations split them 7/7/7.
 * - 42 physical stations provide 2 replicas for each virtual shard.
 *
 * This keeps small meshes sharded instead of falling back to full fan-out.
 */
export function resolveMahaShardRoute(
  stationCount: number,
  prefixHash: string | bigint,
): MahaShardRoute | null {
  if (stationCount <= 0) return null;

  const shard = mapToMahaShard(prefixHash);
  const start = Math.floor((shard * stationCount) / MAHA_ANCHOR);
  const end = Math.floor(((shard + 1) * stationCount) / MAHA_ANCHOR);
  return {
    shard,
    start,
    end: Math.max(start + 1, end),
  };
}

/**
 * Returns the virtual shard range owned by a physical station.
 *
 * Useful for operators exposing a small physical mesh as a stable 21-shard
 * Maha surface. For example, station 0 of a 2-station mesh owns [0, 11),
 * while station 1 owns [11, 21).
 */
export function resolveMahaPhysicalStationRange(
  stationCount: number,
  stationIndex: number,
): MahaPhysicalStationRange | null {
  if (stationCount <= 0) return null;
  if (stationIndex < 0 || stationIndex >= stationCount) return null;

  const shardStart = Math.ceil((stationIndex * MAHA_ANCHOR) / stationCount);
  const shardEnd = Math.ceil(((stationIndex + 1) * MAHA_ANCHOR) / stationCount);
  return {
    stationIndex,
    shardStart,
    shardEnd,
  };
}

/**
 * Selects the subset of stations that should handle a given prefix hash
 * in a sharded cluster.
 *
 * In a "Planetary Scale" deployment, we have MAHA_ANCHOR (21) virtual shards.
 * A physical node might host one or more shards.
 */
export function resolveMahaShardedStations<T>(
  allStations: T[],
  prefixHash: string | bigint
): T[] {
  if (allStations.length === 0) return [];

  const route = resolveMahaShardRoute(allStations.length, prefixHash);
  if (route) return allStations.slice(route.start, route.end);

  return [];
}
