// @a0n/gnosis/distributed-inference-host/src/attractor.ts
// φ-attractor primitives for principled σ selection in the image-carrier
// deblur pipeline.
//
// SSOT note: the deficit-ratio Fibonacci sliding window is `interfere()` /
// `InterfereState` from `@a0n/aether/src/glossolalia-moa.ts:160`. We import
// it directly rather than mirroring the math; the wrappers below
// (`createGoldenAttractorState`, `updateGoldenAttractor`) just thread the
// σ-selection-specific args (no entropy regime tracking, no token absorbing
// state — both irrelevant for picking a Wiener σ) through interfere()'s
// general API.
//
// The PHI literal at glossolalia-moa.ts:200 (`const PHI = 1.618`) remains
// where it is — extracting it across packages is a separate SSOT pass.
// Here we use the exact `(1 + √5) / 2`, which agrees with that literal to
// f64 precision. SQRT_5 is the Binet normalizer in F(n) = (φⁿ − ψⁿ)/√5.

import {
  createInterfereState,
  interfere,
  type InterfereState,
} from '@a0n/aether/src/glossolalia-moa';

export const PHI: number = (1 + Math.sqrt(5)) / 2; // 1.618033988749895
export const PHI_INVERSE: number = 2 / (1 + Math.sqrt(5)); // 0.6180339887498949
export const PHI_SQUARED: number = PHI * PHI; // 2.618033988749895 (= φ + 1)
export const SQRT_5: number = Math.sqrt(5); // 2.23606797749979
export const PSI: number = (1 - Math.sqrt(5)) / 2; // -0.6180339887498949 (Binet conjugate)

/** Closed-form Fibonacci via Binet: F(n) = (φⁿ - ψⁿ) / √5.
 *  Returns the integer-rounded value for n >= 0. F(0)=0, F(1)=1, F(2)=1, ... */
export function binetFibonacci(n: number): number {
  if (!Number.isInteger(n) || n < 0) {
    throw new RangeError(
      `binetFibonacci: n must be a non-negative integer, got ${n}`,
    );
  }
  return Math.round((Math.pow(PHI, n) - Math.pow(PSI, n)) / SQRT_5);
}

/** Distance from the φ-attractor, normalized by √5: (deficitRatio - φ) / √5.
 *  Sign convention: positive means above the attractor (over-correction);
 *  negative means below (under-correction). */
export function binetResidual(deficitRatio: number): number {
  if (!Number.isFinite(deficitRatio)) {
    throw new RangeError(
      `binetResidual: deficitRatio must be finite, got ${deficitRatio}`,
    );
  }
  return (deficitRatio - PHI) / SQRT_5;
}

/** φ-spaced grid: [σ_min·φ⁰, σ_min·φ¹, ..., σ_min·φ^(count-1)].
 *  count must be >= 2 (need at least two points to form a deficit ratio). */
export function goldenSpacedGrid(sigmaMin: number, count: number): number[] {
  if (!Number.isFinite(sigmaMin) || sigmaMin <= 0) {
    throw new RangeError(
      `goldenSpacedGrid: sigmaMin must be > 0, got ${sigmaMin}`,
    );
  }
  if (!Number.isInteger(count) || count < 2) {
    throw new RangeError(
      `goldenSpacedGrid: count must be an integer >= 2, got ${count}`,
    );
  }
  const grid = new Array<number>(count);
  for (let i = 0; i < count; i++) {
    grid[i] = sigmaMin * Math.pow(PHI, i);
  }
  return grid;
}

/** σ-selection alias for `InterfereState`. The underlying type carries
 *  `currentSpread`, `baseSpread`, `entropyHistory`, `repeatCount`,
 *  `lastTokenId` — irrelevant for σ selection but harmless along for the ride. */
export type GoldenAttractorState = InterfereState;

/** Initialize a deficit-ratio sliding window. Wraps `createInterfereState(1.0)`;
 *  `baseSpread` is the temperature-spread reference for MOA decode and is
 *  unused in σ selection. Seeds lastDeficit = prevDeficit = 1 (Fibonacci floor). */
export function createGoldenAttractorState(): GoldenAttractorState {
  return createInterfereState(1.0);
}

/** Slide the Fibonacci window with a new observation. Wraps `interfere()`
 *  with `currentEntropy = 0` (bypasses C2 regime detection — irrelevant for
 *  σ selection) and a synthetic `selectedTokenId = state.cycle` (always
 *  unique → bypasses C3 absorbing-state detection). The returned state's
 *  `deficitRatio` is the only field σ-selection consumers need. */
export function updateGoldenAttractor(
  state: GoldenAttractorState,
  currentDeficit: number,
): GoldenAttractorState {
  if (!Number.isFinite(currentDeficit)) {
    throw new RangeError(
      `updateGoldenAttractor: currentDeficit must be finite, got ${currentDeficit}`,
    );
  }
  return interfere(state, currentDeficit, 0, state.cycle, PHI_INVERSE);
}

/** Linear-interpolation Newton step toward σ where deficitRatio = φ.
 *  Given two adjacent samples (prevSigma, prevRatio) and (sigma, ratio),
 *  estimate σ* = sigma - (ratio - φ) · (sigma - prevSigma) / (ratio - prevRatio).
 *  Clamped to [prevSigma / φ, sigma · φ] — refusing to leap more than one
 *  φ-step. When the slope (ratio - prevRatio) is near zero, returns sigma
 *  unchanged. */
export function phiAttractorNewtonStep(args: {
  sigma: number;
  prevSigma: number;
  ratio: number;
  prevRatio: number;
}): number {
  const { sigma, prevSigma, ratio, prevRatio } = args;
  if (
    !Number.isFinite(sigma) ||
    !Number.isFinite(prevSigma) ||
    !Number.isFinite(ratio) ||
    !Number.isFinite(prevRatio)
  ) {
    throw new RangeError(
      `phiAttractorNewtonStep: all args must be finite, got ${JSON.stringify(args)}`,
    );
  }
  const slope = ratio - prevRatio;
  if (Math.abs(slope) < 1e-12) {
    return sigma;
  }
  const estimate = sigma - ((ratio - PHI) * (sigma - prevSigma)) / slope;
  const lo = prevSigma / PHI;
  const hi = sigma * PHI;
  if (estimate < lo) return lo;
  if (estimate > hi) return hi;
  return estimate;
}
