// @a0n/gnosis/distributed-inference-host/src/audio-denoise.ts
// Spectral subtraction (Boll 1979) — the classical kernel for audio noise
// reduction. This is distinct from audio-deblur.ts, which applies the 2D
// image-deblur math to a spectrogram and is structurally valid but
// mathematically wrong for whisper-quality uplift. Use this module when the
// goal is signal-above-noise recovery (whispers, low-SNR speech). The
// quietest 10% of frames seed the noise template; over-subtraction α
// (default 2.0) controls aggressiveness; spectral floor β (default 0.01)
// prevents "musical noise" artefacts. Caveat: when the audio is loud
// throughout (no near-silence frames), the quietest-10% template will
// contain signal and denoising may attenuate signal too.

import { stft, istft, type STFTConfig } from "./audio-stft";

export interface DenoiseAudioArgs {
  /** Mono audio in [-1, 1]. */
  audio: Float32Array;
  sampleRate: number;
  /** STFT FFT size. Power of 2. Default 1024. */
  nFft?: number;
  /** Hop length. Default nFft / 4. */
  hopLength?: number;
  /** Over-subtraction factor. Default 2.0. Higher = more aggressive (riskier). */
  oversubtractionFactor?: number;
  /** Spectral floor. Default 0.01. Higher = less musical noise but more residual hiss. */
  spectralFloor?: number;
  /** Fraction of frames (lowest total magnitude) used to estimate noise.
   *  Default 0.1 (10%). Caller should ensure the audio contains some near-silence. */
  noiseFractionFrames?: number;
}

export interface DenoiseAudioResult {
  /** Denoised audio in same length range as input (within an STFT frame). */
  audio: Float32Array;
  sampleRate: number;
  nFft: number;
  hopLength: number;
  oversubtractionFactor: number;
  spectralFloor: number;
  /** Estimated noise PSD per frequency bin. Length = nFreqs. */
  noisePsd: Float32Array;
  /** Number of frames used in noise estimation. */
  noiseFrameCount: number;
  /** Mean per-frame magnitude before subtraction (reference for SNR ratio). */
  inputRms: number;
  /** Mean per-frame magnitude after subtraction. */
  outputRms: number;
}

function isPow2(n: number): boolean {
  return Number.isInteger(n) && n > 0 && (n & (n - 1)) === 0;
}

/**
 * Handles the gnosis denoise Audio workflow.
 */
export function denoiseAudio(args: DenoiseAudioArgs): DenoiseAudioResult {
  const {
    audio,
    sampleRate,
    nFft = 1024,
    hopLength = Math.floor((args.nFft ?? 1024) / 4),
    oversubtractionFactor = 2.0,
    spectralFloor = 0.01,
    noiseFractionFrames = 0.1,
  } = args;

  // ---------- validation ----------
  if (!isPow2(nFft)) {
    throw new RangeError(`nFft must be a positive power of 2; got ${nFft}`);
  }
  if (!Number.isFinite(hopLength) || hopLength <= 0 || hopLength > nFft) {
    throw new RangeError(
      `hopLength must be in (0, nFft]; got ${hopLength} with nFft=${nFft}`,
    );
  }
  if (!(audio instanceof Float32Array) || audio.length < nFft) {
    throw new RangeError(
      `audio must be a Float32Array with length >= nFft (${nFft}); got length ${audio?.length}`,
    );
  }
  if (!Number.isFinite(sampleRate) || sampleRate <= 0) {
    throw new RangeError(`sampleRate must be > 0; got ${sampleRate}`);
  }
  if (!Number.isFinite(oversubtractionFactor) || oversubtractionFactor <= 0) {
    throw new RangeError(
      `oversubtractionFactor must be > 0; got ${oversubtractionFactor}`,
    );
  }
  if (
    !Number.isFinite(spectralFloor) ||
    spectralFloor <= 0 ||
    spectralFloor > 1
  ) {
    throw new RangeError(
      `spectralFloor must be in (0, 1]; got ${spectralFloor}`,
    );
  }
  if (
    !Number.isFinite(noiseFractionFrames) ||
    noiseFractionFrames <= 0 ||
    noiseFractionFrames > 1
  ) {
    throw new RangeError(
      `noiseFractionFrames must be in (0, 1]; got ${noiseFractionFrames}`,
    );
  }

  // ---------- forward STFT ----------
  const stftConfig: STFTConfig = { nFft, hopLength };
  const { magnitude, phase, nFrames, nFreqs } = stft(audio, stftConfig);

  // ---------- per-frame total magnitude (for noise-frame selection) ----------
  const rowSums = new Float32Array(nFrames);
  for (let f = 0; f < nFrames; f++) {
    let s = 0;
    const base = f * nFreqs;
    for (let k = 0; k < nFreqs; k++) {
      s += magnitude[base + k]!;
    }
    rowSums[f] = s;
  }

  // ---------- pick quietest fraction of frames ----------
  const indices = new Array<number>(nFrames);
  for (let i = 0; i < nFrames; i++) indices[i] = i;
  indices.sort((a, b) => rowSums[a]! - rowSums[b]!);

  const noiseFrameCount = Math.max(
    1,
    Math.ceil(noiseFractionFrames * nFrames),
  );

  // ---------- noise PSD: mean(|X(k)|²) over selected frames ----------
  const noisePsd = new Float32Array(nFreqs);
  for (let i = 0; i < noiseFrameCount; i++) {
    const f = indices[i]!;
    const base = f * nFreqs;
    for (let k = 0; k < nFreqs; k++) {
      const m = magnitude[base + k]!;
      noisePsd[k]! += m * m;
    }
  }
  for (let k = 0; k < nFreqs; k++) {
    noisePsd[k]! /= noiseFrameCount;
  }

  // ---------- spectral subtraction ----------
  const cleanedMagnitude = new Float32Array(nFrames * nFreqs);
  let inputMagSum = 0;
  let outputMagSum = 0;

  for (let f = 0; f < nFrames; f++) {
    const base = f * nFreqs;
    for (let k = 0; k < nFreqs; k++) {
      const X = magnitude[base + k]!;
      const X2 = X * X;
      const N2 = noisePsd[k]!;
      const Y2 = Math.max(X2 - oversubtractionFactor * N2, spectralFloor * N2);
      const Y = Math.sqrt(Y2);
      cleanedMagnitude[base + k] = Y;
      inputMagSum += X;
      outputMagSum += Y;
    }
  }

  const totalBins = nFrames * nFreqs;
  const inputRms = totalBins > 0 ? inputMagSum / totalBins : 0;
  const outputRms = totalBins > 0 ? outputMagSum / totalBins : 0;

  // ---------- inverse STFT with original phase ----------
  const cleanedAudio = istft({
    magnitude: cleanedMagnitude,
    phase,
    nFrames,
    config: stftConfig,
  });

  return {
    audio: cleanedAudio,
    sampleRate,
    nFft,
    hopLength,
    oversubtractionFactor,
    spectralFloor,
    noisePsd,
    noiseFrameCount,
    inputRms,
    outputRms,
  };
}
