// @a0n/gnosis/distributed-inference-host/src/audio-deblur.ts
// audio-deblur.ts
//
// Audio recovery via 2D image deblur on a log-magnitude spectrogram.
// Pipeline: STFT -> log1p+normalize -> deblurAnyImage (Wiener; phi-attractor
// selects sigma over time-frequency space) -> inverse log (expm1*scale) ->
// Griffin-Lim phase reconstruction -> iSTFT.
//
// **Honest scope note (2026-04-27)**: this is the same image-deblur pipeline
// applied to a magnitude spectrogram. It is NOT a substitute for classical
// audio noise reduction (1D Wiener subtraction, spectral subtraction, RNN
// denoisers). For genuine whisper-amplification — recovering signal that
// sits below a stationary noise floor — the right primitive is spectral
// subtraction with an estimated noise template, not 2D deconvolution. This
// module is the structural slot ("audio is a 2D image-shaped problem") and
// validates the pipeline shape end-to-end; the math for whisper-quality
// uplift will likely need a different inner kernel. Callers should treat
// this as an experimental scaffold, not a finished product.
//
// mel-spectrogram.ts already exists for Whisper-specific configs and is
// intentionally not used here so this module stays config-agnostic.

import { stft, type STFTConfig } from './audio-stft';
import { griffinLim } from './griffin-lim';
import {
  deblurAnyImage,
  type DeblurAnyImageOptions,
} from './deblur-any-image';
import { denoiseAudio } from './audio-denoise';

export interface DeblurAudioArgs {
  /** Mono Float32 audio in [-1, 1]. Length should be at least nFft. */
  audio: Float32Array;
  /** Sample rate (only used in the result for reference; pipeline is sr-agnostic). */
  sampleRate: number;
  /** STFT FFT size. Power of 2. Default 1024. */
  nFft?: number;
  /** Hop length. Default nFft / 4. */
  hopLength?: number;
  /** Griffin-Lim iterations after deblur. Default 30. */
  griffinLimIterations?: number;
  /** Forwarded to deblurAnyImage. `channelMode` is forced to 'luminance' (1ch).
   *  Defaults: sigma='auto', method='wiener', noiseFloor='auto'. */
  options?: DeblurAnyImageOptions;
  /** Pad the spectrogram to power-of-2 dimensions before deblur (deblurAnyImage
   *  mirror-pads internally, but pre-padding the time axis to pow2 reduces the
   *  mirror-pad reach when the spectrogram is far from pow2 in nFrames).
   *  Default true. */
  padTimeAxis?: boolean;
  /** Inner-kernel mode:
   *   - 'image-deblur' (default): apply 2D Wiener/RL on log-magnitude spectrogram.
   *     Structurally valid, not the right math for whisper amplification.
   *   - 'spectral-subtract': Boll 1979 spectral subtraction. Real audio noise
   *     reduction; estimates noise from quietest frames, subtracts.
   *   - 'both': run spectral-subtract first to clean the input, then 2D deblur
   *     for any remaining smoothing benefit. */
  mode?: 'image-deblur' | 'spectral-subtract' | 'both';
  /** Spectral-subtraction tuning, used when mode includes spectral-subtract.
   *  See `audio-denoise.ts` for full details. */
  spectralSubtraction?: {
    oversubtractionFactor?: number;
    spectralFloor?: number;
    noiseFractionFrames?: number;
  };
  /** Initial phase for Griffin-Lim. Default 'zero'. */
  initialPhase?: 'zero' | 'random' | Float32Array;
}

export interface DeblurAudioResult {
  /** Recovered audio (length matches input within an STFT frame's worth of edge). */
  audio: Float32Array;
  sampleRate: number;
  /** sigma selected by the phi-attractor over the spectrogram. */
  selectedSigma: number;
  /** Noise floor used (estimated if 'auto'). */
  noiseFloorUsed: number;
  /** Spectrogram dimensions sent to deblurAnyImage. */
  spectrogramWidth: number;   // nFreqs (always pow2-friendly via deblur's mirrorPad)
  spectrogramHeight: number;  // nFrames (or padded)
  /** Method used by the underlying deblur. */
  methodUsed: 'wiener' | 'richardson-lucy' | 'mesh-residual';
  /** Iteration count for Griffin-Lim. */
  griffinLimIterations: number;
  /** Mode actually used. */
  modeUsed: 'image-deblur' | 'spectral-subtract' | 'both';
  /** Final phase after Griffin-Lim (if used). */
  phase?: Float32Array;
}

function isPowerOfTwo(n: number): boolean {
  return n > 0 && (n & (n - 1)) === 0;
}

/**
 * Handles the gnosis deblur Audio workflow.
 */
export async function deblurAudio(args: DeblurAudioArgs): Promise<DeblurAudioResult> {
  // 1. Validate args.
  const nFft = args.nFft ?? 1024;
  const hopLength = args.hopLength ?? Math.floor(nFft / 4);
  const griffinLimIterations = args.griffinLimIterations ?? 30;

  if (!(args.audio instanceof Float32Array)) {
    throw new RangeError('deblurAudio: audio must be a Float32Array');
  }
  if (!Number.isFinite(args.sampleRate) || args.sampleRate <= 0) {
    throw new RangeError(
      `deblurAudio: sampleRate must be > 0, got ${args.sampleRate}`,
    );
  }
  if (!Number.isInteger(nFft) || nFft <= 0 || !isPowerOfTwo(nFft)) {
    throw new RangeError(
      `deblurAudio: nFft must be a positive power of 2, got ${nFft}`,
    );
  }
  if (!Number.isInteger(hopLength) || hopLength <= 0 || hopLength > nFft) {
    throw new RangeError(
      `deblurAudio: hopLength must be in (0, nFft]=${nFft}, got ${hopLength}`,
    );
  }
  if (args.audio.length < nFft) {
    throw new RangeError(
      `deblurAudio: audio.length (${args.audio.length}) must be >= nFft (${nFft})`,
    );
  }
  if (
    !Number.isInteger(griffinLimIterations) ||
    griffinLimIterations < 0
  ) {
    throw new RangeError(
      `deblurAudio: griffinLimIterations must be a non-negative integer, got ${griffinLimIterations}`,
    );
  }
  const mode = args.mode ?? 'image-deblur';
  if (mode !== 'image-deblur' && mode !== 'spectral-subtract' && mode !== 'both') {
    throw new RangeError(
      `deblurAudio: mode must be one of 'image-deblur' | 'spectral-subtract' | 'both', got ${mode}`,
    );
  }

  // === Mode branch =========================================================
  // 'spectral-subtract' (Boll 1979): pure noise-reduction kernel. Skip the
  // 2D image deblur entirely. This is the right math for whisper-quality
  // signal recovery — see audio-denoise.ts.
  if (mode === 'spectral-subtract') {
    const ssOpts = args.spectralSubtraction ?? {};
    const ss = denoiseAudio({
      audio: args.audio,
      sampleRate: args.sampleRate,
      nFft,
      hopLength,
      oversubtractionFactor: ssOpts.oversubtractionFactor,
      spectralFloor: ssOpts.spectralFloor,
      noiseFractionFrames: ssOpts.noiseFractionFrames,
    });
    return {
      audio: ss.audio,
      sampleRate: args.sampleRate,
      selectedSigma: 0,            // no σ chosen (spectral-subtract bypasses image deblur)
      noiseFloorUsed: ss.spectralFloor, // surface SS floor as the analogous knob
      spectrogramWidth: ss.noisePsd.length,
      spectrogramHeight: 0,
      methodUsed: 'wiener',        // not really used; sentinel
      griffinLimIterations: 0,     // SS has no Griffin-Lim stage
      modeUsed: 'spectral-subtract',
    };
  }

  // 'both': denoise first, then feed the cleaned audio into the image-deblur
  // pipeline. The 2D smoothing on the post-SS spectrogram can suppress
  // residual musical noise that pure SS leaves behind. Costlier than either
  // alone but the cleanest output for hard cases.
  let stagedAudio = args.audio;
  if (mode === 'both') {
    const ssOpts = args.spectralSubtraction ?? {};
    const ss = denoiseAudio({
      audio: args.audio,
      sampleRate: args.sampleRate,
      nFft,
      hopLength,
      oversubtractionFactor: ssOpts.oversubtractionFactor,
      spectralFloor: ssOpts.spectralFloor,
      noiseFractionFrames: ssOpts.noiseFractionFrames,
    });
    stagedAudio = ss.audio;
  }

  // 2. Defaults for deblur options. channelMode is forced to 'luminance'
  //    regardless of caller intent — we always feed a 1-channel spectrogram.
  const baseOptions: DeblurAnyImageOptions = {
    sigma: 'auto',
    method: 'wiener',
    noiseFloor: 'auto',
    ...(args.options ?? {}),
    channelMode: 'luminance',
  } as DeblurAnyImageOptions;

  // NOTE: args.padTimeAxis is a forward-compat hint. For this first cut we do
  // NOT actually pre-pad — deblurAnyImage already mirror-pads (nFreqs, nFrames)
  // internally. The slot exists so a future implementor can pad the time axis
  // to a power of 2 before handing off (reduces mirror-pad reach when nFrames
  // is far from pow2). Leaving this as a documented no-op for now.
  void args.padTimeAxis;

  // 3. Forward STFT — uses stagedAudio so 'both' mode sees the denoised input.
  const stftConfig: STFTConfig = { nFft, hopLength };
  const result = stft(stagedAudio, stftConfig);
  // result.magnitude is Float32Array of length nFrames * nFreqs.

  // 4. Convert magnitude to a [0, 1]-normalized log-magnitude image.
  //    log1p is numerically stable for small values; raw |X| has heavy tails
  //    so the deblur (which expects roughly [0,1] image data) behaves much
  //    better on log-magnitude.
  const N = result.magnitude.length;
  const logMag = new Float32Array(N);
  let maxLog = 0;
  for (let i = 0; i < N; i++) {
    const v = Math.log1p(Math.max(0, result.magnitude[i]));
    logMag[i] = v;
    if (v > maxLog) maxLog = v;
  }
  // Guard against all-zero input (silent audio): logScale=0 collapses the
  // pipeline to a no-op magnitude of 0 everywhere.
  const logScale = maxLog > 0 ? maxLog : 1;
  if (maxLog > 0) {
    const inv = 1 / logScale;
    for (let i = 0; i < N; i++) logMag[i] *= inv;
  }

  // 5. Treat logMag as a 1-channel image of width=nFreqs, height=nFrames.
  //    deblurAnyImage's DecodedImageInput accepts Float32Array directly, so
  //    no Uint8 conversion is needed.

  // 6. Call deblurAnyImage on the spectrogram.
  const deblurred = await deblurAnyImage(
    {
      data: logMag,
      width: result.nFreqs,
      height: result.nFrames,
      channels: 1,
    },
    baseOptions,
  );

  // 7. Convert recovered logMag back to magnitude.
  //    Deblur output is unclamped — clamp to [0,1] before inverse log so
  //    expm1 doesn't blow up, then clip negatives that fall out (Wiener can
  //    produce slightly-negative residuals near the noise floor).
  const cleanLogMag = deblurred.recovered;
  if (cleanLogMag.length !== N) {
    throw new Error(
      `deblurAudio: deblur returned ${cleanLogMag.length} samples, expected ${N}`,
    );
  }
  const cleanMagnitude = new Float32Array(N);
  for (let i = 0; i < N; i++) {
    let x = cleanLogMag[i];
    if (x < 0) x = 0;
    else if (x > 1) x = 1;
    let m = Math.expm1(x * logScale);
    if (m < 0) m = 0;
    cleanMagnitude[i] = m;
  }

  // 8. Griffin-Lim phase recovery -> iSTFT (handled inside griffinLim).
  const gl = griffinLim({
    magnitude: cleanMagnitude,
    nFrames: result.nFrames,
    config: stftConfig,
    iterations: griffinLimIterations,
    initialPhase: args.initialPhase ?? 'zero',
    seed: 0xc0ffee, // deterministic
  });

  // 9. Return the result.
  return {
    audio: gl.audio,
    phase: gl.phase,
    sampleRate: args.sampleRate,
    selectedSigma: deblurred.selectedSigma,
    noiseFloorUsed: deblurred.noiseFloorUsed,
    spectrogramWidth: result.nFreqs,
    spectrogramHeight: result.nFrames,
    methodUsed: deblurred.methodUsed,
    griffinLimIterations,
    modeUsed: mode,
  };
}

/**
 * Enhanced version of deblurAudio that processes long files in chunks to avoid
 * memory exhaustion. Recommended for audio longer than 10 minutes.
 */
export async function deblurAudioLong(
  args: DeblurAudioArgs,
  chunkSizeSeconds = 300, // 5 minutes default
  overlapSeconds = 2, // 2 seconds overlap for cross-fade
): Promise<DeblurAudioResult> {
  const { audio, sampleRate } = args;
  const samplesPerChunk = Math.floor(chunkSizeSeconds * sampleRate);
  const samplesOverlap = Math.floor(overlapSeconds * sampleRate);

  if (audio.length <= samplesPerChunk + samplesOverlap) {
    return deblurAudio(args);
  }

  const results: Float32Array[] = [];
  let offset = 0;
  let lastSigma = 0;
  let lastNoiseFloor = 0;
  let lastPhase: Float32Array | undefined;

  const nFft = args.nFft ?? 1024;
  const hopLength = args.hopLength ?? Math.floor(nFft / 4);
  const nFreqs = nFft / 2 + 1;

  while (offset < audio.length) {
    let end = offset + samplesPerChunk + samplesOverlap;
    if (end > audio.length) end = audio.length;

    if (end - offset < nFft) {
      break; // Remaining audio is too small for a full FFT frame
    }

    const chunk = audio.subarray(offset, end);
    const chunkIdx = Math.floor(offset / samplesPerChunk);
    const totalChunks = Math.ceil(audio.length / samplesPerChunk);
    console.log(`     [Phase B] Mesh Deblur: Processing chunk ${chunkIdx + 1}/${totalChunks}...`);
    
    // Prepare initial phase for this chunk if we have it from the previous overlap
    let initialPhase: 'zero' | Float32Array = 'zero';
    if (lastPhase) {
      const overlapFrames = Math.floor(samplesOverlap / hopLength);
      const totalFramesInChunk = Math.floor((chunk.length - nFft) / hopLength) + 1;
      const phaseToKeep = new Float32Array(totalFramesInChunk * nFreqs);
      
      const lastPhaseFrames = lastPhase.length / nFreqs;
      const startFrame = Math.max(0, lastPhaseFrames - overlapFrames);
      const framesToCopy = Math.min(overlapFrames, totalFramesInChunk);
      
      phaseToKeep.set(lastPhase.subarray(startFrame * nFreqs, (startFrame + framesToCopy) * nFreqs));
      initialPhase = phaseToKeep;
    }

    // Optimization: Reuse sigma from first chunk to avoid expensive attractor loop on every chunk
    const chunkArgs = { ...args, audio: chunk, initialPhase };
    if (lastSigma > 0) {
      chunkArgs.options = { ...args.options, sigma: lastSigma };
    }

    const result = await deblurAudio(chunkArgs);

    const processed = result.audio;
    lastSigma = result.selectedSigma;
    lastNoiseFloor = result.noiseFloorUsed;
    lastPhase = result.phase;

    if (results.length === 0) {
      results.push(processed);
    } else {
      const prevChunk = results[results.length - 1]!;
      const fadeLen = Math.min(samplesOverlap, processed.length, prevChunk.length);
      for (let i = 0; i < fadeLen; i++) {
        const alpha = i / (fadeLen - 1);
        const prevIdx = prevChunk.length - fadeLen + i;
        prevChunk[prevIdx] = prevChunk[prevIdx] * (1 - alpha) + processed[i] * alpha;
      }
      results.push(processed.subarray(fadeLen));
    }

    offset += samplesPerChunk;
    if (offset >= audio.length) break;

    // Manual GC trigger if exposed (helps prevent OOM in long loops with large buffers)
    if (typeof global !== 'undefined' && (global as any).gc) {
      (global as any).gc();
    }
  }

  // Flatten results
  const totalLength = results.reduce((acc, r) => acc + r.length, 0);
  const finalAudio = new Float32Array(totalLength);
  let finalOffset = 0;
  for (const r of results) {
    finalAudio.set(r, finalOffset);
    finalOffset += r.length;
  }

  return {
    audio: finalAudio,
    sampleRate,
    selectedSigma: lastSigma,
    noiseFloorUsed: lastNoiseFloor,
    spectrogramWidth: 0,
    spectrogramHeight: 0,
    methodUsed: 'wiener',
    griffinLimIterations: args.griffinLimIterations ?? 30,
    modeUsed: args.mode ?? 'both',
  };
}
