// @a0n/gnosis/distributed-inference-host/src/betti-signature.ts
// Betti-number sieve. Computes a thresholded super-level-set signature
// (beta0, beta1, beta2) on a 2D Float32 plane and exposes it as a sieve
// criterion for the FRF ensemble. "Pleromatic Closure": sum(beta) <=
// TOWER_CEILINGS.decagon (10) marks the config admissible. Higher sums
// indicate "invented topology" — content the deblur produced that has no
// homological signature in the source neighbourhood. Discrete Euler
// characteristic chi = V - E + F via the 8-connected union-find component
// count + edge/face inventory; beta1 solved from chi = beta0 - beta1 +
// beta2 (open-surface approximation).

export interface BettiSignature {
  beta0: number;       // connected components of {plane >= threshold}
  beta1: number;       // loop count (Euler-derived)
  beta2: number;       // enclosed voids
  euler: number;       // discrete Euler characteristic V - E + F
  total: number;       // beta0 + beta1 + beta2 (the closure norm)
  threshold: number;   // actual threshold value used (= thresholdFraction * max(plane))
}

export interface BettiSignatureArgs {
  plane: Float32Array;
  width: number;
  height: number;
  /** Super-level-set threshold as a fraction of plane max. Default 0.5. */
  thresholdFraction?: number;
  /** Connectivity for component counting. 4 (orthogonal) or 8 (diagonal).
   *  Default 8. */
  connectivity?: 4 | 8;
}

/** Default Pleromatic Closure ceiling — TOWER_CEILINGS.decagon = 10. */
const DEFAULT_CLOSURE_CEILING = 10;

const ZERO_SIGNATURE: BettiSignature = Object.freeze({
  beta0: 0,
  beta1: 0,
  beta2: 0,
  euler: 0,
  total: 0,
  threshold: 0,
});

function validateArgs(args: BettiSignatureArgs): void {
  const { plane, width, height, thresholdFraction, connectivity } = args;
  if (!Number.isInteger(width) || width <= 0) {
    throw new RangeError(`bettiSignature: width must be a positive integer, got ${width}`);
  }
  if (!Number.isInteger(height) || height <= 0) {
    throw new RangeError(`bettiSignature: height must be a positive integer, got ${height}`);
  }
  if (!(plane instanceof Float32Array)) {
    throw new RangeError(`bettiSignature: plane must be a Float32Array`);
  }
  if (plane.length !== width * height) {
    throw new RangeError(
      `bettiSignature: plane.length ${plane.length} does not match width*height ${width * height}`,
    );
  }
  if (thresholdFraction !== undefined) {
    if (
      typeof thresholdFraction !== 'number' ||
      Number.isNaN(thresholdFraction) ||
      thresholdFraction <= 0 ||
      thresholdFraction > 1
    ) {
      throw new RangeError(
        `bettiSignature: thresholdFraction must be in (0, 1], got ${thresholdFraction}`,
      );
    }
  }
  if (connectivity !== undefined && connectivity !== 4 && connectivity !== 8) {
    throw new RangeError(`bettiSignature: connectivity must be 4 or 8, got ${connectivity}`);
  }
}

/** Disjoint-set union with path compression and union-by-rank. */
class DSU {
  private parent: Int32Array;
  private rank: Uint8Array;

  constructor(n: number) {
    this.parent = new Int32Array(n);
    this.rank = new Uint8Array(n);
    for (let i = 0; i < n; i++) this.parent[i] = i;
  }

  find(x: number): number {
    let r = x;
    while (this.parent[r] !== r) r = this.parent[r];
    // path compression
    let cur = x;
    while (this.parent[cur] !== r) {
      const next = this.parent[cur];
      this.parent[cur] = r;
      cur = next;
    }
    return r;
  }

  union(a: number, b: number): void {
    const ra = this.find(a);
    const rb = this.find(b);
    if (ra === rb) return;
    if (this.rank[ra] < this.rank[rb]) {
      this.parent[ra] = rb;
    } else if (this.rank[ra] > this.rank[rb]) {
      this.parent[rb] = ra;
    } else {
      this.parent[rb] = ra;
      this.rank[ra]++;
    }
  }
}

/**
 * Handles the gnosis betti Signature workflow.
 */
export function bettiSignature(args: BettiSignatureArgs): BettiSignature {
  validateArgs(args);
  const { plane, width, height } = args;
  const thresholdFraction = args.thresholdFraction ?? 0.5;
  const connectivity: 4 | 8 = args.connectivity ?? 8;

  // Find max
  let max = -Infinity;
  for (let i = 0; i < plane.length; i++) {
    const v = plane[i];
    if (v > max) max = v;
  }

  // All-zero (or all non-positive) input → return zeros
  if (!Number.isFinite(max) || max <= 0) {
    return { ...ZERO_SIGNATURE };
  }

  const threshold = thresholdFraction * max;

  // Build binary mask
  const N = width * height;
  const mask = new Uint8Array(N);
  let V = 0;
  for (let i = 0; i < N; i++) {
    if (plane[i] >= threshold) {
      mask[i] = 1;
      V++;
    }
  }

  if (V === 0) {
    return { beta0: 0, beta1: 0, beta2: 0, euler: 0, total: 0, threshold };
  }

  // β₀ via union-find
  const dsu = new DSU(N);
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      const i = y * width + x;
      if (!mask[i]) continue;
      // neighbours: right, down, plus diagonals if 8-conn
      // (we only need to check forward neighbours to avoid double-union)
      if (x + 1 < width && mask[i + 1]) dsu.union(i, i + 1);
      if (y + 1 < height && mask[i + width]) dsu.union(i, i + width);
      if (connectivity === 8) {
        if (x + 1 < width && y + 1 < height && mask[i + width + 1]) {
          dsu.union(i, i + width + 1);
        }
        if (x - 1 >= 0 && y + 1 < height && mask[i + width - 1]) {
          dsu.union(i, i + width - 1);
        }
      }
    }
  }

  const rootSet = new Set<number>();
  for (let i = 0; i < N; i++) {
    if (mask[i]) rootSet.add(dsu.find(i));
  }
  const beta0 = rootSet.size;

  // Discrete Euler characteristic χ = V - E + F
  // V counted above. E = 4-connected adjacent mask pairs.
  // F = 2x2 squares fully inside mask.
  let E = 0;
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      const i = y * width + x;
      if (!mask[i]) continue;
      if (x + 1 < width && mask[i + 1]) E++;
      if (y + 1 < height && mask[i + width]) E++;
    }
  }

  let F = 0;
  for (let y = 0; y < height - 1; y++) {
    for (let x = 0; x < width - 1; x++) {
      const i = y * width + x;
      if (mask[i] && mask[i + 1] && mask[i + width] && mask[i + width + 1]) F++;
    }
  }

  const euler = V - E + F;

  // β₂ via flood-fill on complement from the border
  // visited[i]=1 means cell is mask=0 AND reachable from the image border.
  const visited = new Uint8Array(N);
  // Stack-based flood fill (4-connectivity for the void definition: anything
  // diagonally pinched off is still considered enclosed).
  const stack: number[] = [];
  // Seed with all border mask=0 cells
  for (let x = 0; x < width; x++) {
    const top = x;
    const bot = (height - 1) * width + x;
    if (!mask[top] && !visited[top]) {
      visited[top] = 1;
      stack.push(top);
    }
    if (!mask[bot] && !visited[bot]) {
      visited[bot] = 1;
      stack.push(bot);
    }
  }
  for (let y = 0; y < height; y++) {
    const left = y * width;
    const right = y * width + (width - 1);
    if (!mask[left] && !visited[left]) {
      visited[left] = 1;
      stack.push(left);
    }
    if (!mask[right] && !visited[right]) {
      visited[right] = 1;
      stack.push(right);
    }
  }

  while (stack.length > 0) {
    const i = stack.pop()!;
    const x = i % width;
    const y = (i - x) / width;
    // 4-connected neighbours
    if (x + 1 < width) {
      const j = i + 1;
      if (!mask[j] && !visited[j]) {
        visited[j] = 1;
        stack.push(j);
      }
    }
    if (x - 1 >= 0) {
      const j = i - 1;
      if (!mask[j] && !visited[j]) {
        visited[j] = 1;
        stack.push(j);
      }
    }
    if (y + 1 < height) {
      const j = i + width;
      if (!mask[j] && !visited[j]) {
        visited[j] = 1;
        stack.push(j);
      }
    }
    if (y - 1 >= 0) {
      const j = i - width;
      if (!mask[j] && !visited[j]) {
        visited[j] = 1;
        stack.push(j);
      }
    }
  }

  // Now count connected components of unreached mask=0 cells (the voids).
  // 4-connected union-find on the void set (consistent with the flood).
  const voidDsu = new DSU(N);
  let voidCount = 0;
  for (let i = 0; i < N; i++) {
    if (!mask[i] && !visited[i]) voidCount++;
  }
  if (voidCount === 0) {
    // β₂ = 0
  }
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      const i = y * width + x;
      if (mask[i] || visited[i]) continue;
      if (x + 1 < width) {
        const j = i + 1;
        if (!mask[j] && !visited[j]) voidDsu.union(i, j);
      }
      if (y + 1 < height) {
        const j = i + width;
        if (!mask[j] && !visited[j]) voidDsu.union(i, j);
      }
    }
  }
  const voidRoots = new Set<number>();
  for (let i = 0; i < N; i++) {
    if (!mask[i] && !visited[i]) voidRoots.add(voidDsu.find(i));
  }
  const beta2 = voidRoots.size;

  // β₁ from χ = β₀ - β₁ + β₂  ⇒  β₁ = β₀ + β₂ - χ
  let beta1 = beta0 + beta2 - euler;
  if (beta1 < 0) beta1 = 0;

  const total = beta0 + beta1 + beta2;

  return { beta0, beta1, beta2, euler, total, threshold };
}

/** Median across a set of signatures, used as the FRF ensemble reference.
 *  Returns the per-coordinate median (NOT the magnitude — use Manhattan
 *  distance for sieve scoring). */
export function medianSignature(
  sigs: ReadonlyArray<BettiSignature>,
): { beta0: number; beta1: number; beta2: number } {
  if (sigs.length === 0) return { beta0: 0, beta1: 0, beta2: 0 };
  const b0 = sigs.map((s) => s.beta0).sort((a, b) => a - b);
  const b1 = sigs.map((s) => s.beta1).sort((a, b) => a - b);
  const b2 = sigs.map((s) => s.beta2).sort((a, b) => a - b);
  const pickMedian = (arr: number[]): number => {
    const n = arr.length;
    if (n % 2 === 1) return arr[(n - 1) >> 1];
    return (arr[n / 2 - 1] + arr[n / 2]) / 2;
  };
  return {
    beta0: pickMedian(b0),
    beta1: pickMedian(b1),
    beta2: pickMedian(b2),
  };
}

/** Manhattan distance from a signature to a reference point. Used as a
 *  rejection score: higher = more topologically deviant. */
export function signatureDistance(
  sig: BettiSignature,
  ref: { beta0: number; beta1: number; beta2: number },
): number {
  return (
    Math.abs(sig.beta0 - ref.beta0) +
    Math.abs(sig.beta1 - ref.beta1) +
    Math.abs(sig.beta2 - ref.beta2)
  );
}

/** Pleromatic Closure check: sum(beta) <= ceiling. Default ceiling 10
 *  (TOWER_CEILINGS.decagon). Returns true if admissible. */
export function isClosureAdmissible(
  sig: BettiSignature,
  ceiling: number = DEFAULT_CLOSURE_CEILING,
): boolean {
  return Math.abs(sig.beta0) + Math.abs(sig.beta1) + Math.abs(sig.beta2) <= ceiling;
}
