// @a0n/gnosis/distributed-inference-host/src/boson-skyrms.ts
/**
 * # Boson Skyrms Equilibria — runtime mirror
 *
 * Mirrors `Gnosis.BosonSkyrmsEquilibria` (Lean) and the Rust
 * `distributed_inference::cost_algebra` boson surface.
 *
 * The Standard Model's 13 bosons map to specific phase cycles in the
 * `BraidedTower`, with a `topologicallySafe` Skyrms-equilibrium
 * witness for each. The mapping is *signature*, not *derivation* —
 * by `every_phase_count_is_a_tower`, the calculus permits any phase
 * count.
 *
 * The **Higgs ↔ vacuum** identity is the only mapping with a formal
 * payoff: only the Higgs (= vacuum) admits free duplication via the
 * no-cloning theorem, matching the Higgs field's role of giving mass
 * everywhere as a vacuum-broadcast.
 *
 * @see open-source/gnosis-math/Gnosis/BosonSkyrmsEquilibria.lean
 * @see open-source/gnosis/distributed-inference/src/cost_algebra.rs
 */

import { TOWER_CEILINGS, type BuleyProjection } from './spectral-noise-monitor';

/** A Standard Model boson. */
export type StandardModelBoson =
  | { kind: 'photon' }
  | { kind: 'wPlus' }
  | { kind: 'wMinus' }
  | { kind: 'zZero' }
  | { kind: 'gluon'; index: number /* 0..7 */ }
  | { kind: 'higgs' };

/** All 13 Standard Model bosons in canonical order. */
export const allStandardModelBosons: readonly StandardModelBoson[] = [
  { kind: 'photon' },
  { kind: 'wPlus' },
  { kind: 'wMinus' },
  { kind: 'zZero' },
  { kind: 'gluon', index: 0 },
  { kind: 'gluon', index: 1 },
  { kind: 'gluon', index: 2 },
  { kind: 'gluon', index: 3 },
  { kind: 'gluon', index: 4 },
  { kind: 'gluon', index: 5 },
  { kind: 'gluon', index: 6 },
  { kind: 'gluon', index: 7 },
  { kind: 'higgs' },
];

/** The phase cycle (manifold ceiling) each boson occupies. */
export function bosonPhaseCount(boson: StandardModelBoson): number {
  switch (boson.kind) {
    case 'photon':
    case 'wPlus':
    case 'wMinus':
    case 'zZero':
      return TOWER_CEILINGS.triton; // 3 — SU(2)×U(1) electroweak
    case 'gluon':
      return 8; // Octagon — SU(3) adjoint, 8 generators
    case 'higgs':
      return 0; // Vacuum carrier
  }
}

/** A Bule unit sits at a boson's Skyrms equilibrium when its score
 *  exactly equals the boson's phase ceiling (`topologicallySafe`). */
export function bosonSkyrmsEquilibrium(
  b: BuleyProjection,
  boson: StandardModelBoson,
): boolean {
  return b.score === bosonPhaseCount(boson);
}

/** Classify a Bule unit into the boson(s) it currently inhabits the
 *  Skyrms equilibrium of. Returns the matching list — typically 0 or
 *  several (4 electroweak share Triton, 8 gluons share Octagon, the
 *  Higgs is unique at the vacuum). */
export function classifyBosonEquilibrium(b: BuleyProjection): readonly StandardModelBoson[] {
  return allStandardModelBosons.filter((boson) =>
    bosonSkyrmsEquilibrium(b, boson),
  );
}

/** The Higgs ↔ vacuum identity: the Higgs equilibrium iff the
 *  carrier's score is zero. The unique boson whose Skyrms equilibrium
 *  is the vacuum carrier — the only state that admits free fan-out
 *  by the no-cloning theorem. */
export function isHiggsEquilibrium(b: BuleyProjection): boolean {
  return b.score === 0;
}
