// @a0n/gnosis/distributed-inference-host/src/buleyean-rejection-weights.ts
// Rejection-aware Buleyean weighting.
// In the canonical FRF ensemble (forkraceFoldDeblur.ts), rejected candidates
// are filtered by the ceiling and discarded. This module uses their sigma
// coordinates as a prior on the survivors: a survivor close (in sigma-space)
// to a rejected candidate gets its Buleyean rank weight attenuated by a
// proximity factor. Reifies "perfect failure = perfect success" from
// open-source/buleyean-rl: failure is information, not waste, and the
// information density per rejection is log2(N-1) bits in an N-fork run.
// All weights floor at 1 (the sliver) so no candidate is permanently
// silenced — preserves the buleyean-positivity law.

import { buleyeanWeight } from '@a0n/buleyean-kernel';

export interface RejectionPoint {
  /** sigma value of the rejected candidate. */
  sigma: number;
  /** Optional |Binet residual| (or other distance-to-attractor) for finer
   *  weighting; if absent, just sigma-distance is used. */
  attractorDistance?: number;
}

export interface SurvivingCandidate {
  /** sigma value of the surviving candidate. */
  sigma: number;
  /** 0-based rank by ascending |Binet residual| among survivors. */
  rank: number;
}

export interface RejectionWeightArgs {
  survivors: ReadonlyArray<SurvivingCandidate>;
  rejected: ReadonlyArray<RejectionPoint>;
  /** Influence strength of rejection proximity. 0 = ignore rejections (pure
   *  Buleyean rank weight); 1 = full attenuation at the rejection boundary.
   *  Default 0.5. */
  alpha?: number;
}

export interface RejectionWeightResult {
  /** One weight per survivor, in input order. Floored at 1. */
  weights: ReadonlyArray<number>;
  /** Per-survivor diagnostic: closest rejection sigma-distance. */
  proximityDistances: ReadonlyArray<number>;
  /** Total weight (sum of weights). */
  totalWeight: number;
}

/**
 * Fold rejected candidates into survivor weighting via sigma-proximity prior.
 *
 * For each surviving candidate s:
 *   - R = survivors.length
 *   - w_rank = buleyeanWeight(R, s.rank)
 *   - d_s = min |sigma_s - sigma_r| over r in rejected (Infinity if rejected is empty)
 *   - p_s = 1 / (1 + d_s) ∈ (0, 1]
 *   - w_aug = w_rank * (1 - alpha * p_s)
 *   - weights[i] = max(1, w_aug)   // sliver floor
 */
export function rejectionWeightedFold(args: RejectionWeightArgs): RejectionWeightResult {
  const { survivors, rejected } = args;
  const alpha = args.alpha ?? 0.5;

  if (typeof alpha !== 'number' || !Number.isFinite(alpha)) {
    throw new TypeError(`rejectionWeightedFold: alpha must be a finite number, got ${String(alpha)}`);
  }
  if (alpha < 0 || alpha > 1) {
    throw new RangeError(`rejectionWeightedFold: alpha must be in [0, 1], got ${alpha}`);
  }
  if (!survivors || survivors.length < 1) {
    throw new RangeError('rejectionWeightedFold: survivors must contain at least one candidate');
  }

  const R = survivors.length;
  const weights: number[] = new Array(R);
  const proximityDistances: number[] = new Array(R);
  let totalWeight = 0;

  for (let i = 0; i < R; i++) {
    const s = survivors[i];
    const wRank = buleyeanWeight(R, s.rank);

    let d: number;
    if (rejected.length === 0) {
      d = Infinity;
      proximityDistances[i] = d;
      const w = Math.max(1, wRank);
      weights[i] = w;
      totalWeight += w;
      continue;
    }

    let minD = Infinity;
    for (let j = 0; j < rejected.length; j++) {
      const dist = Math.abs(s.sigma - rejected[j].sigma);
      if (dist < minD) minD = dist;
    }
    d = minD;
    proximityDistances[i] = d;

    const p = 1 / (1 + d);
    const wAug = wRank * (1 - alpha * p);
    const w = Math.max(1, wAug);
    weights[i] = w;
    totalWeight += w;
  }

  return {
    weights,
    proximityDistances,
    totalWeight,
  };
}
