// @a0n/gnosis/distributed-inference-host/src/carrier-medium.ts
/**
 * Carrier medium — observer-relative evolution/resolution over typed cells.
 *
 * The medium core is intentionally domain-agnostic: callers provide unit-valued
 * cells, and this module reports how each cell's +1 carrier step is read by
 * observer bandwidths. Images, audio, text, code, and latent grids should adapt
 * into this surface instead of redefining the bandwidth arithmetic.
 */

import {
  BANDWIDTH_HEXON,
  BANDWIDTH_TRIHEXENNEON,
  BANDWIDTH_TRITON,
  buleStepCost,
  isEvolutionStep,
} from './pleromatic-bandwidth';

export type CarrierMediumReadingKind = 'evolution' | 'resolution';

export interface CarrierMediumCellInput<Metadata = unknown> {
  readonly unitValue: number;
  readonly id?: string;
  readonly metadata?: Metadata;
}

export interface CarrierMediumInput<Metadata = unknown> {
  readonly cells: readonly CarrierMediumCellInput<Metadata>[];
}

export interface CarrierMediumPipelineConfig {
  readonly sourceBandwidth?: number;
  readonly observerBandwidths?: readonly number[];
}

export interface CarrierFiber {
  readonly sourceBandwidth: number;
  readonly observerBandwidth: number;
  readonly lowerPosition: number;
  readonly start: number;
  readonly endExclusive: number;
  readonly multiplicity: number;
}

export interface CarrierMediumReading {
  readonly observerBandwidth: number;
  readonly reading: CarrierMediumReadingKind;
  readonly buleCost: 1;
  readonly lowerPosition: number;
  readonly copyIndex: number;
  readonly collisionMultiplicity: number;
  readonly copyIndexRepresentable: boolean;
  readonly silenceOfUnderBandwidth: boolean;
  readonly fiber: CarrierFiber;
}

export interface CarrierMediumCell<Metadata = unknown> {
  readonly index: number;
  readonly id?: string;
  readonly metadata?: Metadata;
  readonly unitValue: number;
  readonly stepOrigin: number;
  readonly nextPosition: number;
  readonly readings: readonly CarrierMediumReading[];
}

export interface CarrierMediumBandwidthSummary {
  readonly observerBandwidth: number;
  readonly evolutionCount: number;
  readonly resolutionCount: number;
  readonly totalBuleCost: number;
  readonly copyIndexGapCount: number;
  readonly maxCollisionMultiplicity: number;
}

export interface CarrierMediumAnalysis<Metadata = unknown> {
  readonly sourceBandwidth: number;
  readonly observerBandwidths: readonly number[];
  readonly cells: readonly CarrierMediumCell<Metadata>[];
  readonly summaries: readonly CarrierMediumBandwidthSummary[];
  readonly buleCostInvariantAcrossBandwidths: boolean;
}

interface MutableSummary {
  observerBandwidth: number;
  evolutionCount: number;
  resolutionCount: number;
  totalBuleCost: number;
  copyIndexGapCount: number;
  maxCollisionMultiplicity: number;
}

const DEFAULT_OBSERVER_BANDWIDTHS = [
  BANDWIDTH_TRITON,
  BANDWIDTH_HEXON,
  BANDWIDTH_TRIHEXENNEON,
] as const;

export function analyzeCarrierMedium<Metadata = unknown>(
  input: CarrierMediumInput<Metadata>,
  config: CarrierMediumPipelineConfig = {},
): CarrierMediumAnalysis<Metadata> {
  return new CarrierMediumPipeline(config).analyze(input);
}

/**
 * Handles the gnosis quantize Unit To Carrier Step workflow.
 */
export function quantizeUnitToCarrierStep(value: number, bandwidth: number): number {
  assertPositiveInteger(bandwidth, 'bandwidth');
  const clamped = clamp01(value);
  return Math.min(bandwidth - 1, Math.floor(clamped * bandwidth));
}

/**
 * Handles the gnosis project Carrier Position workflow.
 */
export function projectCarrierPosition(
  position: number,
  sourceBandwidth: number,
  observerBandwidth: number,
): number {
  assertCarrierPosition(position, sourceBandwidth, 'position');
  assertPositiveInteger(observerBandwidth, 'observerBandwidth');
  return Math.min(
    observerBandwidth - 1,
    Math.floor((position * observerBandwidth) / sourceBandwidth),
  );
}

/**
 * Handles the gnosis carrier Fiber For Position workflow.
 */
export function carrierFiberForPosition(
  position: number,
  sourceBandwidth: number,
  observerBandwidth: number,
): CarrierFiber {
  const lowerPosition = projectCarrierPosition(
    position,
    sourceBandwidth,
    observerBandwidth,
  );
  const start = Math.ceil((lowerPosition * sourceBandwidth) / observerBandwidth);
  const endExclusive = Math.ceil(
    ((lowerPosition + 1) * sourceBandwidth) / observerBandwidth,
  );
  return {
    sourceBandwidth,
    observerBandwidth,
    lowerPosition,
    start,
    endExclusive,
    multiplicity: Math.max(0, endExclusive - start),
  };
}

/**
 * Handles the gnosis classify Carrier Step workflow.
 */
export function classifyCarrierStep(
  stepOrigin: number,
  sourceBandwidth: number,
  observerBandwidth: number,
): CarrierMediumReading {
  assertCarrierPosition(stepOrigin, sourceBandwidth, 'stepOrigin');
  assertPositiveInteger(observerBandwidth, 'observerBandwidth');

  const fiber = carrierFiberForPosition(
    stepOrigin,
    sourceBandwidth,
    observerBandwidth,
  );
  const copyIndex = stepOrigin - fiber.start;
  const copyIndexRepresentable = fiber.multiplicity <= observerBandwidth;
  const reading = isEvolutionStep(observerBandwidth, stepOrigin)
    ? 'evolution'
    : 'resolution';

  return {
    observerBandwidth,
    reading,
    buleCost: buleStepCost(observerBandwidth, stepOrigin),
    lowerPosition: fiber.lowerPosition,
    copyIndex,
    collisionMultiplicity: fiber.multiplicity,
    copyIndexRepresentable,
    silenceOfUnderBandwidth: fiber.multiplicity > 1 && !copyIndexRepresentable,
    fiber,
  };
}

export class CarrierMediumPipeline {
  private readonly sourceBandwidth: number;
  private readonly observerBandwidths: readonly number[];

  constructor(config: CarrierMediumPipelineConfig = {}) {
    const sourceBandwidth = config.sourceBandwidth ?? BANDWIDTH_TRIHEXENNEON;
    assertPositiveInteger(sourceBandwidth, 'sourceBandwidth');

    const observerBandwidths =
      config.observerBandwidths ?? DEFAULT_OBSERVER_BANDWIDTHS;
    if (!observerBandwidths.length) {
      throw new Error('CarrierMediumPipeline: observerBandwidths must be non-empty');
    }
    for (const bandwidth of observerBandwidths) {
      assertPositiveInteger(bandwidth, 'observerBandwidth');
    }

    this.sourceBandwidth = sourceBandwidth;
    this.observerBandwidths = Array.from(observerBandwidths);
  }

  analyze<Metadata = unknown>(
    input: CarrierMediumInput<Metadata>,
  ): CarrierMediumAnalysis<Metadata> {
    const summaries = new Map<number, MutableSummary>();
    for (const bandwidth of this.observerBandwidths) {
      summaries.set(bandwidth, {
        observerBandwidth: bandwidth,
        evolutionCount: 0,
        resolutionCount: 0,
        totalBuleCost: 0,
        copyIndexGapCount: 0,
        maxCollisionMultiplicity: 0,
      });
    }

    const cells = input.cells.map((cellInput, index) => {
      const unitValue = normalizeUnitValue(cellInput.unitValue, index);
      const stepOrigin = quantizeUnitToCarrierStep(
        unitValue,
        this.sourceBandwidth,
      );
      const readings = this.observerBandwidths.map((observerBandwidth) => {
        const reading = classifyCarrierStep(
          stepOrigin,
          this.sourceBandwidth,
          observerBandwidth,
        );
        const summary = summaries.get(observerBandwidth);
        if (!summary) {
          throw new Error(
            `CarrierMediumPipeline: missing summary for bandwidth ${observerBandwidth}`,
          );
        }
        if (reading.reading === 'evolution') {
          summary.evolutionCount += 1;
        } else {
          summary.resolutionCount += 1;
        }
        summary.totalBuleCost += reading.buleCost;
        if (reading.silenceOfUnderBandwidth) {
          summary.copyIndexGapCount += 1;
        }
        summary.maxCollisionMultiplicity = Math.max(
          summary.maxCollisionMultiplicity,
          reading.collisionMultiplicity,
        );
        return reading;
      });

      return {
        index,
        ...(cellInput.id !== undefined ? { id: cellInput.id } : {}),
        ...(cellInput.metadata !== undefined ? { metadata: cellInput.metadata } : {}),
        unitValue,
        stepOrigin,
        nextPosition: stepOrigin + 1,
        readings,
      };
    });

    const frozenSummaries = Array.from(summaries.values()).map((summary) => ({
      ...summary,
    }));
    const expectedCost = frozenSummaries[0]?.totalBuleCost ?? 0;
    const buleCostInvariantAcrossBandwidths = frozenSummaries.every(
      (summary) => summary.totalBuleCost === expectedCost,
    );

    return {
      sourceBandwidth: this.sourceBandwidth,
      observerBandwidths: this.observerBandwidths,
      cells,
      summaries: frozenSummaries,
      buleCostInvariantAcrossBandwidths,
    };
  }
}

function normalizeUnitValue(value: number, index: number): number {
  if (!Number.isFinite(value)) {
    throw new RangeError(`cells[${index}].unitValue must be finite`);
  }
  return clamp01(value);
}

function assertCarrierPosition(position: number, bandwidth: number, label: string): number {
  assertPositiveInteger(bandwidth, 'bandwidth');
  if (!Number.isInteger(position) || position < 0 || position >= bandwidth) {
    throw new RangeError(
      `${label} must be an integer in [0, ${bandwidth})`,
    );
  }
  return position;
}

function assertPositiveInteger(value: number, label: string): number {
  if (!Number.isInteger(value) || value <= 0) {
    throw new RangeError(`${label} must be a positive integer`);
  }
  return value;
}

function clamp01(value: number): number {
  if (value <= 0) return 0;
  if (value >= 1) return 1;
  return value;
}
