// @a0n/gnosis/distributed-inference-host/src/cluster.ts
import { Pipeline, type PipelineConfig } from './pipeline';
import { RegistryShim, type RegistryShimConfig } from './registry-shim';
import { Station as RealStation } from './station';
import { StationUds } from './station-uds';
import type { IStation } from './types';

export interface ClusterConfig {
  /** The knot_id to run inference against. */
  knotId: string;
  /** Hidden dim of the model. */
  hiddenDim: number;
  /** Vocab size of the model. */
  vocabSize?: number;
  /** The URL of the Resonance Registry. */
  registryUrl: string;
  /** 
   * The layer ranges that make up the pipeline.
   * e.g. [[0, 8], [8, 16], [16, 24]]
   */
  layerRanges: [number, number][];
  /** Optional override for pipeline chunk size. */
  chunkSize?: number;
  /** Optional override for bizarro speculation mode. */
  enableBizarro?: boolean;
  /** Optional Maha-anchored sharding for Amplituhedron replay/capture. */
  amplituhedronSharding?: boolean;
  /** Optional custom fetcher for the registry shim. */
  fetcher?: RegistryShimConfig['fetcher'];
  /**
   * TQFT Cobordism vocabulary sharding ranges.
   * Each pair is [vocabStart, vocabEnd). When present, the cluster
   * replaces the standard /embed call with a parallel fan-out to
   * N cobordism stations, merging via the Frobenius identity
   * (μ: A ⊗ A → A — elementwise addition of disjoint partial embeddings).
   *
   * See: FORMAL_LEDGER.md § Atiyah-Segal TQFT Functor
   */
  cobordismRanges?: [number, number][];
  /** Optional cobordism station endpoints (parallel to cobordismRanges). */
  cobordismFetcher?: RegistryShimConfig['fetcher'];
  /** Optional UDS socket paths parallel to layerRanges. */
  udsPaths?: string[];
  /** Optional UDS socket paths parallel to cobordismRanges. */
  cobordismUdsPaths?: string[];
}

/**
 * Operator Facade for a distributed inference cluster.
 * 
 * Assembles a Pipeline by querying the Resonance Registry for Stations
 * across a set of layer ranges. Serves as the "one giant machine" entry point.
 */
export class DistributedInferenceCluster {
  private readonly pipeline: Pipeline;

  private constructor(pipeline: Pipeline) {
    this.pipeline = pipeline;
  }

  /**
   * Open a connection to a distributed inference cluster.
   * Internally builds the log-rolling pipeline using RegistryShim for failover.
   */
  static open(config: ClusterConfig): DistributedInferenceCluster {
    if (!config.layerRanges || config.layerRanges.length === 0) {
      throw new Error('DistributedInferenceCluster: layerRanges must be non-empty');
    }

    const stations: IStation[] = config.layerRanges.map(([startLayer, endLayer], idx) => {
      const isEntry = idx === 0;
      const isExit = idx === config.layerRanges.length - 1;
      const role = isEntry && isExit ? undefined : isEntry ? 'entry' : isExit ? 'exit' : 'mid';

      const udsPath = config.udsPaths?.[idx];
      if (udsPath) {
        return new StationUds({
          socketPath: udsPath,
          startLayer,
          endLayer,
          role,
        });
      }

      const shim = new RegistryShim({
        registryUrl: config.registryUrl,
        knotId: config.knotId,
        startLayer,
        endLayer,
        fetcher: config.fetcher,
      });

      return new RealStation({
        endpoint: shim,
        startLayer,
        endLayer,
        role,
        baseUrl: `http://station-${startLayer}`,
      });
    });

    const pipelineConfig: PipelineConfig = {
      stations,
      hiddenDim: config.hiddenDim,
      vocabSize: config.vocabSize,
      chunkSize: config.chunkSize,
      enableBizarro: config.enableBizarro,
      amplituhedronSharding: config.amplituhedronSharding,
    };

    // TQFT Cobordism: when vocabulary sharding ranges are present,
    // build cobordism stations and attach them to the pipeline config.
    if (config.cobordismRanges && config.cobordismRanges.length > 0) {
      pipelineConfig.cobordismStations = config.cobordismRanges.map(
        ([vocabStart, vocabEnd], idx) => {
          const udsPath = config.cobordismUdsPaths?.[idx];
          if (udsPath) {
            return {
              station: new StationUds({
                socketPath: udsPath,
                startLayer: 0,
                endLayer: 0,
                role: 'cobordism',
              }),
              vocabStart,
              vocabEnd,
            };
          }

          const shim = new RegistryShim({
            registryUrl: config.registryUrl,
            knotId: config.knotId,
            startLayer: 0,
            endLayer: 0,
            fetcher: config.cobordismFetcher ?? config.fetcher,
          });
          return {
            station: new RealStation({
              endpoint: shim,
              startLayer: 0,
              endLayer: 0,
              role: 'cobordism',
              baseUrl: `http://cobordism-${idx}`,
            }),
            vocabStart,
            vocabEnd,
          };
        },
      );
    }

    const pipeline = new Pipeline(pipelineConfig);
    return new DistributedInferenceCluster(pipeline);
  }

  /**
   * Generate tokens autoregressively.
   * 
   * @param promptTokens The initial prompt token ids.
   * @param maxNewTokens The maximum number of tokens to generate.
   * @returns The generated token ids, including the prompt.
   */
  async generate(promptTokens: Uint32Array, maxNewTokens: number): Promise<Uint32Array> {
    return await this.pipeline.generate({ promptTokens, maxNewTokens });
  }

  /**
   * Expose the underlying Pipeline for low-level operations (e.g. prefill only).
   */
  getPipeline(): Pipeline {
    return this.pipeline;
  }
}
