// @a0n/gnosis/distributed-inference-host/src/continuous-batch-scheduler.ts
// Continuous-batch decode scheduler — the read-bound capture's serving layer.
//
// `decode_step_batch` (native, parity-verified) advances K sequences by ONE token each
// in a single pass: every layer's weight is read ONCE for the whole wave (the
// amortization that earns the transformer's read-bound ~6.2×). This scheduler drives it:
// admit sequences into a fixed-width wave, run one `decodeStepBatch` per tick, then fan
// out the per-lane sample+stream. The batched native call IS the polyglot
// `Fanout { n, op, combiner }` (one op over n lanes, native-combined); the TS-side
// per-lane streaming lowers that Fanout to `Promise.all`.
//
// Replaces `withPipelineMutex`'s serialize-everything path when MOONSHINE_CONTINUOUS_BATCH
// is on (default OFF → today's behavior is byte-identical). v1 is STATIC batching: a freed
// lane is not re-filled mid-run (true continuous re-use needs a per-lane KV reset on the
// native binding — flagged below). The math win (one read per wave) is already realized.

/** The native binding surface (NlpHandle: `initBatch` / `resetLane` / `decodeStepBatch`). */
export interface DecodeBatchBinding {
  /** Allocate `n` persistent per-sequence KV slabs. */
  initBatch(n: number): void;
  /** Clear lane `k`'s KV slab + position so it can host a NEW sequence — the per-lane
   *  reset that makes lane RE-USE (true continuous batching) correct, not just static. */
  resetLane(k: number): void;
  /** One batched decode step over `n` lanes: `tokens[k]` is lane k's input token; returns
   *  `n * vocabSize` row-major f32 logits (row k = lane k's next-token logits). */
  decodeStepBatch(tokens: Uint32Array): Float32Array;
}

/** A request occupying one lane of the wave. */
export interface BatchSeq {
  id: string;
  promptTokens: number[];
  maxNewTokens: number;
  /** logits view → next token id (caller owns temp/top-p; greedy for the gate test). */
  sample: (logits: Float32Array, offset: number, vocabSize: number) => number;
  isStop: (token: number) => boolean;
  /** Stream one generated token (may be async — SSE write). */
  onToken: (token: number) => void | Promise<void>;
  onDone: (reason: 'eos' | 'maxTokens') => void;
}

interface Lane {
  seq: BatchSeq;
  feedCount: number; // tokens consumed (prompt then generated)
  generated: number;
  lastToken: number;
  done: boolean;
}

export class ContinuousBatchScheduler {
  private lanes: (Lane | null)[];
  private pending: BatchSeq[] = [];

  constructor(
    private readonly binding: DecodeBatchBinding,
    private readonly vocabSize: number,
    private readonly capacity: number,
  ) {
    if (capacity <= 0) throw new Error('ContinuousBatchScheduler: capacity must be > 0');
    this.binding.initBatch(capacity);
    this.lanes = new Array(capacity).fill(null);
  }

  /** Rough number of free lanes (for a pipelineQueueDepth-style UX). */
  freeLanes(): number {
    return this.lanes.reduce((n, l) => n + (l === null ? 1 : 0), 0);
  }

  /** Sequences waiting for a lane (a pipelineQueueDepth-style backlog). */
  queueDepth(): number {
    return this.pending.length;
  }

  /** Admit a sequence — continuous batching: it joins the running wave as soon as a lane
   *  frees (no head-of-line block), so the wave stays saturated. Always accepted (queued). */
  admit(seq: BatchSeq): void {
    this.pending.push(seq);
  }

  /** Pull waiting sequences into any free lanes, resetting each lane's KV first so the new
   *  sequence starts from a clean slab. This is what makes lane RE-USE correct. */
  private fill(): void {
    for (let k = 0; k < this.capacity && this.pending.length > 0; k++) {
      if (this.lanes[k] === null) {
        this.binding.resetLane(k);
        this.lanes[k] = { seq: this.pending.shift()!, feedCount: 0, generated: 0, lastToken: 0, done: false };
      }
    }
  }

  private nextInput(lane: Lane): number {
    return lane.feedCount < lane.seq.promptTokens.length
      ? lane.seq.promptTokens[lane.feedCount]
      : lane.lastToken;
  }

  /** One batched decode tick. First admits any waiting sequences into freed lanes (the
   *  continuous re-use), then runs one batched step over every live lane. Returns true
   *  while there is work (live lanes or a backlog). */
  async tick(): Promise<boolean> {
    this.fill();
    const anyLive = this.lanes.some((l) => l !== null && !l.done);
    if (!anyLive) return false;

    // Gather one input token per lane (idle/empty lanes feed a 0 — their output is ignored).
    const tokens = new Uint32Array(this.capacity);
    for (let k = 0; k < this.capacity; k++) {
      const lane = this.lanes[k];
      tokens[k] = lane && !lane.done ? this.nextInput(lane) : 0;
    }

    // THE FANOUT: one weight-read serves the whole wave (decode_step_batch).
    const logits = this.binding.decodeStepBatch(tokens);

    // Fold: fan the per-lane sample + stream (the Fanout lowered to Promise.all).
    await Promise.all(
      this.lanes.map(async (lane, k) => {
        if (!lane || lane.done) return;
        lane.feedCount++;
        // Still inside the prompt → this step only seeds KV; nothing to emit yet.
        if (lane.feedCount < lane.seq.promptTokens.length) return;
        const tok = lane.seq.sample(logits, k * this.vocabSize, this.vocabSize);
        lane.lastToken = tok;
        lane.generated++;
        await lane.seq.onToken(tok);
        if (lane.seq.isStop(tok)) {
          lane.done = true;
          lane.seq.onDone('eos');
          this.lanes[k] = null; // v1: freed, not re-filled (true continuous re-use needs KV reset)
        } else if (lane.generated >= lane.seq.maxNewTokens) {
          lane.done = true;
          lane.seq.onDone('maxTokens');
          this.lanes[k] = null;
        }
      }),
    );
    return true;
  }

  /** Drive ticks until every admitted sequence retires. */
  async run(): Promise<void> {
    let guard = 0;
    while (await this.tick()) {
      if (++guard > 1_000_000) throw new Error('ContinuousBatchScheduler: tick guard tripped');
    }
  }
}
