// @a0n/gnosis/distributed-inference-host/src/cost-algebra-broadcast.ts
/**
 * # Broadcast primitives — vacuum-free vs Landauer-charged
 *
 * The cost algebra forces a choice at every fan-out: either the
 * payload's carrier is the vacuum (in which case duplication is free
 * by `pillarFutureAnchor`), or it carries positive Bule score (in which
 * case duplication must pay the no-cloning tax via
 * `requireBuleHotReplicaAdmission`).
 *
 * This module exposes the two paths as separate functions so the
 * call-site choice is explicit and irreducible:
 *
 * - [`vacuumBroadcast`] — only accepts vacuum carriers. Uses
 *   `requirePillarFuture` as the gate. The fan-out generates zero
 *   replication entropy by construction. Ideal for: configuration
 *   updates, immutable model-weight distribution, idempotent heartbeats.
 *
 * - [`costlyBroadcast`] — accepts non-vacuum carriers but charges the
 *   `(k - 1) · score(s)` Landauer tax. Uses
 *   `requireBuleHotReplicaAdmission` as the gate. Rejects when the tax
 *   exceeds the ceiling. Ideal for: stateful replication, Hot-Replica
 *   spawning, consensus rounds.
 *
 * No third path exists. By the no-cloning theorem, attempting to
 * broadcast a non-vacuum carrier "for free" is a CostHom violation.
 *
 * @see ./cost-algebra-enforcement.ts
 * @see open-source/gnosis-math/Gnosis/CostAlgebraNoCloning.lean
 * @see open-source/gnosis-math/Gnosis/LandauerHotReplica.lean
 */

import { buleyCostAlgebra } from './cost-algebra';
import {
  replicationEntropy,
  type HotReplicaAdmission,
} from './cost-algebra-entropy';
import {
  CostAlgebraViolation,
  requireBuleHotReplicaAdmission,
  requirePillarFuture,
} from './cost-algebra-enforcement';
import {
  TOWER_CEILINGS,
  type BuleyProjection,
} from './spectral-noise-monitor';

/** Request to broadcast `payload` to multiple recipients with a given
 *  Bule carrier shape. The carrier represents the operational cost of
 *  the *transmission* itself, not the payload's semantic content. */
export interface VacuumBroadcastRequest<T> {
  /** The payload to fan out. The cost-algebra is silent on its
   *  semantic content — it gates only on the carrier. */
  payload: T;
  /** The Bule projection of the transmission cost. Must be the
   *  vacuum (`{waste:0, opportunity:0, diversity:0}`) to pass the
   *  future-pillar gate. */
  carrier: BuleyProjection;
  /** Recipients of the broadcast. Used for fan-out reporting only. */
  recipients: readonly string[];
}

/** Result of a successful vacuum broadcast. */
export interface VacuumBroadcastResult<T> {
  payload: T;
  fanout: number;
  cost: 0;
}

/** Vacuum broadcast: refuses to run on any carrier with positive
 *  score. The future-pillar gate fires here — by the no-cloning
 *  theorem, only the vacuum admits free duplication, so this is the
 *  only honest fan-out primitive that genuinely costs nothing. */
export function vacuumBroadcast<T>(
  req: VacuumBroadcastRequest<T>,
): VacuumBroadcastResult<T> {
  requirePillarFuture(req.carrier);
  return {
    payload: req.payload,
    fanout: req.recipients.length,
    cost: 0,
  };
}

/** Request to broadcast under the Landauer ceiling, paying the
 *  no-cloning tax. Use this when the carrier carries genuine
 *  operational cost. */
export interface CostlyBroadcastRequest<T> {
  payload: T;
  carrier: BuleyProjection;
  recipients: readonly string[];
  /** Heat budget the source can absorb before melting. Defaults to
   *  the two-head Hexon ceiling (6). */
  ceiling?: number;
}

/** Result of a successful costly broadcast: includes the admission
 *  decision and the entropy paid for the fan-out. */
export interface CostlyBroadcastResult<T> {
  payload: T;
  admission: HotReplicaAdmission<BuleyProjection>;
  /** Entropy paid by the fan-out — `(k - 1) · score(carrier)`. */
  replicationEntropy: number;
}

/** Costly broadcast: charges the Landauer tax for fanning out a
 *  non-vacuum carrier. Throws `CostAlgebraViolation` if the total heat
 *  exceeds the ceiling. */
export function costlyBroadcast<T>(
  req: CostlyBroadcastRequest<T>,
): CostlyBroadcastResult<T> {
  const ceiling = req.ceiling ?? TOWER_CEILINGS.hexon;
  const k = req.recipients.length;
  const admission = requireBuleHotReplicaAdmission(req.carrier, k, ceiling);
  const entropy = replicationEntropy(buleyCostAlgebra, req.carrier, k);
  return {
    payload: req.payload,
    admission,
    replicationEntropy: entropy,
  };
}

/** Choose the right broadcast primitive based on the carrier's score.
 *  Returns a discriminated union so the caller can branch on the
 *  cost path. Useful when broadcast call-sites can't statically know
 *  whether their carrier is vacuum or not. */
export type BroadcastResult<T> =
  | { kind: 'vacuum'; result: VacuumBroadcastResult<T> }
  | { kind: 'costly'; result: CostlyBroadcastResult<T> };

export function smartBroadcast<T>(
  payload: T,
  carrier: BuleyProjection,
  recipients: readonly string[],
  ceiling: number = TOWER_CEILINGS.hexon,
): BroadcastResult<T> {
  if (carrier.score === 0) {
    return {
      kind: 'vacuum',
      result: vacuumBroadcast({ payload, carrier, recipients }),
    };
  }
  return {
    kind: 'costly',
    result: costlyBroadcast({ payload, carrier, recipients, ceiling }),
  };
}

export { CostAlgebraViolation };
