// @a0n/gnosis/distributed-inference-host/src/cost-algebra-entropy.ts
/**
 * # Entropy + No-Cloning + Hot-Replica admission — runtime core
 *
 * Mirrors the Lean modules:
 *
 * - `Gnosis.CostAlgebraNoCloning` — the diagonal `Δ : S → S × S` is a
 *   `CostHom` iff score is identically zero. In the runtime, this
 *   manifests as: only `vacuumBuleyProjection` admits free duplication.
 * - `Gnosis.CostAlgebraEntropy` — cloning generates exactly `score(s)`
 *   of entropy; replication to k total copies costs `(k-1) · score(s)`.
 * - `Gnosis.LandauerHotReplica` — admission predicate
 *   `k · score(s) ≤ ceiling`, with deficit and remediation guidance.
 *
 * The runtime uses these to:
 *
 * - quantify the cost of replicating a layer-node's state for high
 *   availability;
 * - reject Hot-Replica requests that would melt the source;
 * - emit deterministic remediation (corrective contracts) when
 *   admission fails.
 *
 * @see open-source/gnosis-math/Gnosis/CostAlgebraNoCloning.lean
 * @see open-source/gnosis-math/Gnosis/CostAlgebraEntropy.lean
 * @see open-source/gnosis-math/Gnosis/LandauerHotReplica.lean
 */

import { Verdict, isAccept, isDefer, isReject } from '@a0n/bitwise/triton-verdict';

import type { CostAlgebra } from './cost-algebra';
import { buleyCostAlgebra } from './cost-algebra';
import type { BuleyProjection } from './spectral-noise-monitor';

export { Verdict, isAccept, isDefer, isReject };

/** The diagonal map `Δ : S → S × S`, `s ↦ (s, s)`. Always a function on
 *  carriers; not always a CostHom (see `isDiagonalScorePreserving`). */
export function diagonal<S>(s: S): readonly [S, S] {
  return [s, s] as const;
}

/** Score of the cloned pair under the product cost algebra: it is
 *  always `score s + score s`. The entropy generated by cloning is the
 *  difference, which equals `score s` (the "no-free-lunch tax"). */
export function clonedPairScore<S>(A: CostAlgebra<S>, s: S): number {
  return A.score(s) + A.score(s);
}

/** The entropy generated by cloning `s`: equal to `A.score(s)`.
 *  Mirrors `entropyGeneratedByCloning` from
 *  `Gnosis.CostAlgebraEntropy`. */
export function entropyGeneratedByCloning<S>(A: CostAlgebra<S>, s: S): number {
  return A.score(s);
}

/** Predicate: is the diagonal score-preserving on `s`?  True iff
 *  `A.score(s) === 0`, i.e. `s` is the vacuum (or otherwise scoreless).
 *  Mirrors `diagonal_preserves_score_at`. */
export function isDiagonalScorePreserving<S>(A: CostAlgebra<S>, s: S): boolean {
  return A.score(s) === 0;
}

/** Replication entropy: replicating `s` to `k ≥ 1` total replicas
 *  generates `(k-1) · score(s)` of entropy. For `k = 0` the entropy is
 *  zero (no replicas). Mirrors
 *  `Gnosis.CostAlgebraEntropy.replicationEntropy`. */
export function replicationEntropy<S>(
  A: CostAlgebra<S>,
  s: S,
  k: number,
): number {
  if (k <= 0) return 0;
  return (k - 1) * A.score(s);
}

/** Total heat carried by a `k`-replica fan-out: source's score plus
 *  generated entropy. For `k ≥ 1` this is `k · score(s)`. Mirrors
 *  `Gnosis.LandauerHotReplica.replicaTotalHeat`. */
export function replicaTotalHeat<S>(
  A: CostAlgebra<S>,
  s: S,
  k: number,
): number {
  if (k <= 0) return A.score(s);
  return A.score(s) + replicationEntropy(A, s, k);
}

/**
 * Default multiplier from the soft admit `ceiling` to the hard thermal
 * limit. Heat in `(ceiling, ceiling * THERMAL_HARD_LIMIT_FACTOR]` is the
 * "thermal middle": over the soft ceiling, but not yet structurally
 * unsafe — the right action is admit-with-throttle/backpressure rather
 * than a flat reject. Above the hard limit the source would melt and the
 * request is structurally undeniable (Decline). `2.0` ⇒ the throttle band
 * is exactly one ceiling wide above the ceiling.
 */
export const THERMAL_HARD_LIMIT_FACTOR = 2.0;

/** A Hot-Replica admission decision. */
export interface HotReplicaAdmission<S> {
  /**
   * Whether the request is admissible under the (soft) ceiling. This is
   * the LOSSY collapse of {@link verdict}: `admitted === isAccept(verdict)`,
   * i.e. `totalHeat <= ceiling`. Both `Abstain` (throttle) and `Decline`
   * (melt) collapse to `admitted === false`, which is exactly the old
   * binary behavior — the bool cannot tell throttle from reject.
   */
  admitted: boolean;
  /**
   * The canonical ternary thermal verdict — the distinction the bool
   * `admitted` cannot express:
   *   - `Accept`  — `totalHeat <= ceiling`: comfortably admissible.
   *   - `Abstain` — `ceiling < totalHeat <= hardLimit`: the thermal middle.
   *     Over the soft ceiling but not structurally unsafe; the right action
   *     is admit-with-throttle / backpressure (defer), not a flat reject.
   *   - `Decline` — `totalHeat > hardLimit`: would melt the source; reject.
   */
  verdict: Verdict;
  /** Total heat the request would generate (source + replication entropy). */
  totalHeat: number;
  /** Heat budget the node can absorb before melting (soft admit ceiling). */
  ceiling: number;
  /**
   * Hard thermal limit: above this the source is structurally unsafe and
   * the request is Declined. The band `(ceiling, hardLimit]` is the
   * throttle/Abstain region.
   */
  hardLimit: number;
  /** Excess over the (soft) ceiling, or 0 if admitted. */
  deficit: number;
  /** Source state's own score. */
  sourceScore: number;
  /** Number of total replicas requested. */
  replicas: number;
}

/** Admission decision for replicating `s` to `k` total replicas under a
 *  Landauer thermal `ceiling`. Mirrors
 *  `Gnosis.LandauerHotReplica.admitsReplication` and
 *  `landauerDeficit`.
 *
 *  Now a natural 3-way thermal admission via {@link Verdict}: comfortably
 *  under the soft `ceiling` ⇒ `Accept`; over the structurally-unsafe hard
 *  limit ⇒ `Decline`; in the thermal middle band `(ceiling, hardLimit]`,
 *  where the right action is admit-with-throttle/backpressure ⇒ `Abstain`
 *  (defer). The legacy `admitted` boolean is retained as the lossy collapse
 *  `isAccept(verdict)` — it cannot distinguish throttle from melt. The hard
 *  limit defaults to `ceiling * THERMAL_HARD_LIMIT_FACTOR`; pass an explicit
 *  `hardLimit` to size the throttle band directly (a `hardLimit === ceiling`
 *  removes the band entirely, faithfully degrading to a 2-way
 *  Accept/Decline split). */
export function hotReplicaAdmission<S>(
  A: CostAlgebra<S>,
  s: S,
  k: number,
  ceiling: number,
  hardLimit: number = ceiling * THERMAL_HARD_LIMIT_FACTOR,
): HotReplicaAdmission<S> {
  const totalHeat = replicaTotalHeat(A, s, k);
  // Guard against an inverted band: the hard limit can never sit below the
  // soft ceiling (that would make the throttle region negative-width).
  const effectiveHardLimit = Math.max(hardLimit, ceiling);
  const admitted = totalHeat <= ceiling;
  const verdict: Verdict = admitted
    ? Verdict.Accept
    : totalHeat <= effectiveHardLimit
      ? Verdict.Abstain // thermal middle: over soft ceiling, admit-with-throttle
      : Verdict.Decline; // over the hard limit: structurally unsafe, would melt
  const deficit = admitted ? 0 : totalHeat - ceiling;
  return {
    admitted,
    verdict,
    totalHeat,
    ceiling,
    hardLimit: effectiveHardLimit,
    deficit,
    sourceScore: A.score(s),
    replicas: k,
  };
}

/** Bule-specialized admission check — the common case for the
 *  inference mesh. */
export function buleHotReplicaAdmission(
  s: BuleyProjection,
  k: number,
  ceiling: number,
  hardLimit?: number,
): HotReplicaAdmission<BuleyProjection> {
  return hotReplicaAdmission(buleyCostAlgebra, s, k, ceiling, hardLimit);
}

/** Round-up corrective-contract count: how many `clinamenContract`
 *  steps the source needs to take so that `k · (score - n) ≤ ceiling`.
 *  Returns 0 when already admitted. */
export function correctiveContractsForReplica<S>(
  A: CostAlgebra<S>,
  s: S,
  k: number,
  ceiling: number,
): number {
  const decision = hotReplicaAdmission(A, s, k, ceiling);
  if (decision.admitted) return 0;
  if (k <= 0) return 0;
  // We need k * (score - n) ≤ ceiling, so n ≥ score - floor(ceiling / k).
  const sourceScore = A.score(s);
  const allowedScore = Math.floor(ceiling / k);
  return Math.max(0, sourceScore - allowedScore);
}
