// @a0n/gnosis/distributed-inference-host/src/cost-algebra-enforcement.ts
/**
 * # Runtime enforcement of cost-algebra constraints
 *
 * `cost-algebra.ts` and `cost-algebra-entropy.ts` expose advisory
 * predicates: they classify candidate states and report whether they
 * conform to the algebra. This module turns those advisors into
 * **gates**: throwing variants that refuse to proceed on violation.
 *
 * Two complementary use modes:
 *
 * - **strict** — any violation throws `CostAlgebraViolation`. The
 *   no-cloning theorem and the manifold ceiling become hard runtime
 *   invariants.
 * - **advisory** — caller invokes the predicates from `cost-algebra*`
 *   directly and decides what to do with the result. The default for
 *   diagnostic tooling.
 *
 * There is no third "universal-accept" mode: the algebra of cost has
 * one zero element (the vacuum) where everything is admissible, and
 * routing all states to the vacuum is the same as not running the
 * mesh. So strict and advisory are the only honest choices.
 *
 * @see ./cost-algebra.ts
 * @see ./cost-algebra-entropy.ts
 * @see ./pillar-anchors.ts
 */

import type { CostAlgebra } from './cost-algebra';
import { buleyCostAlgebra } from './cost-algebra';
import {
  hotReplicaAdmission,
  type HotReplicaAdmission,
} from './cost-algebra-entropy';
import {
  pillarFutureAnchor,
  pillarPastAnchor,
  pillarPresentAnchor,
  threePillarReport,
  type ThreePillarReport,
} from './pillar-anchors';
import {
  TOWER_CEILINGS,
  type BuleyProjection,
} from './spectral-noise-monitor';

/** A violation of a cost-algebra invariant at the runtime gate. */
export class CostAlgebraViolation extends Error {
  /** Which gate the violation tripped. */
  readonly gate:
    | 'hot-replica-admission'
    | 'cost-hom'
    | 'pillar-past'
    | 'pillar-present'
    | 'pillar-future'
    | 'three-pillar';
  /** Optional structured context: the decision/report that triggered the throw. */
  readonly context: unknown;

  constructor(
    gate: CostAlgebraViolation['gate'],
    message: string,
    context: unknown = null,
  ) {
    super(message);
    this.name = 'CostAlgebraViolation';
    this.gate = gate;
    this.context = context;
  }
}

/** Strict variant of `hotReplicaAdmission`: throws on rejection. The
 *  source state is unchanged on success; the caller should apply
 *  `correctiveContractsForReplica` before retrying on failure. */
export function requireHotReplicaAdmission<S>(
  A: CostAlgebra<S>,
  s: S,
  k: number,
  ceiling: number,
): HotReplicaAdmission<S> {
  const decision = hotReplicaAdmission(A, s, k, ceiling);
  if (!decision.admitted) {
    throw new CostAlgebraViolation(
      'hot-replica-admission',
      `Hot Replica rejected: total heat ${decision.totalHeat} exceeds ceiling ${decision.ceiling} by ${decision.deficit}`,
      decision,
    );
  }
  return decision;
}

/** Bule-specialized strict admission for the inference mesh. */
export function requireBuleHotReplicaAdmission(
  s: BuleyProjection,
  k: number,
  ceiling: number = TOWER_CEILINGS.hexon,
): HotReplicaAdmission<BuleyProjection> {
  return requireHotReplicaAdmission(buleyCostAlgebra, s, k, ceiling);
}

/** Strict variant of `validateCostHom`: throws if the candidate map
 *  fails any law on the witness set. Returns nothing on success. */
export function requireCostHom<S, T>(
  candidate: (s: S) => T,
  source: CostAlgebra<S>,
  target: CostAlgebra<T>,
  witnesses: readonly S[],
): void {
  const failures: string[] = [];

  if (!target.equals(candidate(source.zero), target.zero)) {
    failures.push('preserve_zero');
  }

  for (const a of witnesses) {
    for (const b of witnesses) {
      const lhs = candidate(source.compose(a, b));
      const rhs = target.compose(candidate(a), candidate(b));
      if (!target.equals(lhs, rhs)) {
        failures.push(`preserve_compose@(${source.score(a)},${source.score(b)})`);
        break;
      }
    }
  }

  for (const s of witnesses) {
    if (target.score(candidate(s)) !== source.score(s)) {
      failures.push(`preserve_score@${source.score(s)}`);
    }
  }

  if (failures.length > 0) {
    throw new CostAlgebraViolation(
      'cost-hom',
      `Candidate map failed cost-hom laws: ${failures.join(', ')}`,
      { failures },
    );
  }
}

/** Strict variant of `pillarPastAnchor`: throws if the state isn't
 *  reachable from the vacuum. (In practice, every Bule unit is — this
 *  gate exists for completeness with the other pillars.) */
export function requirePillarPast(b: BuleyProjection): void {
  if (!pillarPastAnchor(b)) {
    throw new CostAlgebraViolation(
      'pillar-past',
      `Pillar I (Past) anchor failed: state not reachable from vacuum`,
      { state: b },
    );
  }
}

/** Strict variant of `pillarPresentAnchor`: throws if the state is
 *  outside the manifold. */
export function requirePillarPresent(
  b: BuleyProjection,
  ceiling: number = TOWER_CEILINGS.hexon,
): void {
  const result = pillarPresentAnchor(b, ceiling);
  if (!result.insideManifold) {
    throw new CostAlgebraViolation(
      'pillar-present',
      `Pillar II (Present) anchor failed: state score ${result.violation?.score} exceeds manifold ceiling ${result.ceiling} by ${result.violation?.correctiveContractCount}`,
      result,
    );
  }
}

/** Strict variant of `pillarFutureAnchor`: throws if the state is not
 *  free-duplicable. By the no-cloning theorem, this passes only at
 *  the vacuum. **Use sparingly** — most operational states will fail
 *  this gate, and that is correct behavior. The gate is for code paths
 *  that explicitly require duplication-without-cost (e.g. broadcast
 *  primitives that should refuse to run on non-vacuum carriers). */
export function requirePillarFuture(b: BuleyProjection): void {
  const result = pillarFutureAnchor(b);
  if (!result.freeDuplication) {
    throw new CostAlgebraViolation(
      'pillar-future',
      `Pillar III (Future) anchor failed: state with score ${b.score} cannot be duplicated freely (would generate ${result.cloningEntropy} entropy)`,
      result,
    );
  }
}

/** Strict variant of `threePillarReport`: throws if **any** pillar
 *  fails. Returns the full report on success. The Past pillar is
 *  always anchored; only Present and Future can fail in practice. */
export function requireThreePillars(
  b: BuleyProjection,
  ceiling: number = TOWER_CEILINGS.hexon,
): ThreePillarReport {
  const report = threePillarReport(b, ceiling);
  const failed: string[] = [];
  if (!report.past.anchored) failed.push('past');
  if (!report.present.insideManifold) failed.push('present');
  if (!report.future.freeDuplication) failed.push('future');
  if (failed.length > 0) {
    throw new CostAlgebraViolation(
      'three-pillar',
      `Three-pillar gate failed on: ${failed.join(', ')}`,
      report,
    );
  }
  return report;
}

/** Strict-but-vacuum-tolerant: like `requireThreePillars`, but allows
 *  the future pillar to fail (since most non-vacuum states will). The
 *  practical operational gate: past + present must hold; future is
 *  reported but not enforced. */
export function requireOperationalPillars(
  b: BuleyProjection,
  ceiling: number = TOWER_CEILINGS.hexon,
): ThreePillarReport {
  const report = threePillarReport(b, ceiling);
  const failed: string[] = [];
  if (!report.past.anchored) failed.push('past');
  if (!report.present.insideManifold) failed.push('present');
  if (failed.length > 0) {
    throw new CostAlgebraViolation(
      'three-pillar',
      `Operational gate (past + present) failed on: ${failed.join(', ')}`,
      report,
    );
  }
  return report;
}
