// @a0n/gnosis/distributed-inference-host/src/cross-node-coincidence.ts
/**
 * # Cross-Node Coincidence — quorum admission, single-ear refutation, spoofer rejection
 *
 * Mirrors `Gnosis.Mesh.CrossNodeCoincidence` from `open-source/gnosis-math`,
 * the physics-supplied generalization of the BGM2Pose single-ear hard negative
 * (cf. `./active-acoustic-pose`). Where the single-ear acoustic gate needed a
 * *calibration* hard negative (a same-source/different-pose pair recorded on
 * purpose) to refute "maybe source identity is doing the work", a mesh of ears
 * gets its hard negative from **geometry for free**: several independent ears
 * hearing the same event from different vantage points produce arrival times
 * that are jointly consistent with one source only when the event is real. A
 * single ear's hallucination (sensor noise, an internal artifact, a spoofed
 * injection) cannot fabricate arrival times at the *other* nodes that line up
 * with a common source.
 *
 * ## What is (and is not) carried
 *
 * This module is the runtime promotion check for a cross-node acoustic
 * coincidence. It carries ONLY the finite contract the Lean file proves:
 *
 *  - model the predicted arrival at a node from a source hypothesis
 *    (`predictedArrival`: emission time plus 1-D corridor distance);
 *  - threshold a node's arrival residual against a tolerance (`consistent`);
 *  - count the mutually-consistent ears (`countConsistent`);
 *  - admit only when a quorum `K` of ears agree, and never at `K < 2` — a lone
 *    ear can never self-certify (`single_ear_not_admitted`);
 *  - reject a geometrically-inconsistent spoofer so it cannot inflate the
 *    coincidence count (`spoofer_rejected` / `spoofer_does_not_inflate`).
 *
 * True clock synchronization across nodes, Euclidean distance in 2-D/3-D (here
 * distance is the 1-D corridor metric), real propagation speed, multipath, and
 * TDoA solver accuracy stay runtime/empirical obligations OUTSIDE this
 * certificate — they are not theorems and are not asserted here.
 *
 * The promotion check folds the contract into the same epistemic-outcome
 * vocabulary the sibling acoustic carrier already speaks
 * (`claim_admitted` / `observation_only` / `claim_withheld`), so a detection
 * cannot be displayed as an event claim until a quorum of independent ears has
 * cleared the Lean-side geometry/quorum gate.
 *
 * ## Lean source-of-truth
 *
 * @see open-source/gnosis-math/Gnosis/Mesh/CrossNodeCoincidence.lean
 * @see open-source/gnosis-math/Gnosis/Mesh/ActiveAcousticPose.lean (single-ear sibling)
 */

/* ------------------------------------------------------------------ */
/*  Nodes, hypotheses, and arrivals                                   */
/* ------------------------------------------------------------------ */

/**
 * A mesh ear: where it sits on the 1-D corridor, the tick it reported hearing
 * the event, and whether it detected anything at all. Positions and arrivals
 * are non-negative integers (saturating `Nat` arithmetic), matching the Lean
 * `Node`.
 */
export interface MeshNode {
  readonly position: number;
  readonly reportedArrival: number;
  readonly detected: boolean;
}

/** A candidate explanation: a source at a position emitting at a tick.
 *  Mirrors `CrossNodeCoincidence.SourceHypothesis`. */
export interface SourceHypothesis {
  readonly position: number;
  readonly emissionTime: number;
}

/** Absolute difference on `Nat` via saturating subtraction (`|a - b|`). The
 *  1-D corridor distance, and the residual metric for arrivals.
 *  Mirrors `CrossNodeCoincidence.adist`. */
export function adist(a: number, b: number): number {
  return Math.max(0, a - b) + Math.max(0, b - a);
}

/** Predicted arrival at a position under a hypothesis: emission time plus the
 *  distance the carrier travels (speed normalized to one unit per tick).
 *  Mirrors `CrossNodeCoincidence.predictedArrival`. */
export function predictedArrival(h: SourceHypothesis, pos: number): number {
  return h.emissionTime + adist(h.position, pos);
}

/** The arrival residual: how far a node's reported arrival is from the arrival
 *  the hypothesis predicts at that node.
 *  Mirrors `CrossNodeCoincidence.residual`. */
export function residual(h: SourceHypothesis, node: MeshNode): number {
  return adist(node.reportedArrival, predictedArrival(h, node.position));
}

/* ------------------------------------------------------------------ */
/*  Consistency and quorum admission                                  */
/* ------------------------------------------------------------------ */

/** A node is consistent with a hypothesis when it detected the event and its
 *  arrival residual clears the tolerance.
 *  Mirrors `CrossNodeCoincidence.consistentB`. */
export function consistent(
  tol: number,
  h: SourceHypothesis,
  node: MeshNode,
): boolean {
  return node.detected && residual(h, node) <= tol;
}

/** How many nodes in a list are consistent with the hypothesis.
 *  Mirrors `CrossNodeCoincidence.countConsistent`. */
export function countConsistent(
  tol: number,
  h: SourceHypothesis,
  nodes: readonly MeshNode[],
): number {
  let count = 0;
  for (const node of nodes) {
    if (consistent(tol, h, node)) {
      count += 1;
    }
  }
  return count;
}

/* ------------------------------------------------------------------ */
/*  Runtime promotion check (epistemic outcome)                       */
/* ------------------------------------------------------------------ */

/**
 * Epistemic outcome of a cross-node coincidence observation. Identical
 * vocabulary to the sibling acoustic carrier (`./active-acoustic-pose`), kept
 * as a local literal union so this certificate has no import coupling:
 *
 *  - `claim_admitted`    — a quorum `K >= 2` of independent ears are mutually
 *                          consistent with one source hypothesis, so the
 *                          coincidence may be promoted to a real-event claim.
 *  - `observation_only`  — one or more ears are consistent but the quorum is
 *                          not met (an unquorate, real-looking detection), OR
 *                          the requested quorum is `K < 2` and so cannot
 *                          self-certify; the observation is reported, never
 *                          promoted.
 *  - `claim_withheld`    — no ear is consistent with the hypothesis, so there
 *                          is no detection to claim.
 */
export type CrossNodeEpistemicOutcome =
  | 'claim_admitted'
  | 'observation_only'
  | 'claim_withheld';

export interface CrossNodeCoincidenceAdmission {
  readonly outcome: CrossNodeEpistemicOutcome;
  /** Indices (into the node list) of the ears consistent with the hypothesis. */
  readonly consistentNodes: number[];
  /** How many ears are mutually consistent with the hypothesis. */
  readonly coincidenceCount: number;
  /** Whether the quorum was met (count >= K AND K >= 2). */
  readonly quorumMet: boolean;
  readonly summary: string;
}

/**
 * The promotion check Monster Mesh runs before any UI surface displays a
 * cross-node acoustic coincidence as a real event. It folds the Lean contract
 * into an epistemic outcome: the coincidence count decides whether there is
 * anything to claim, and the quorum `K >= 2` decides whether the claim is
 * allowed to be a *real-event* claim rather than the report of a lone ear that
 * cannot self-certify.
 *
 * A quorum of `K < 2` is refused outright (mirrors how the Lean module forbids
 * quorum-1 admission via `single_ear_not_admitted`): a single ear — real,
 * noisy, or spoofed — cannot manufacture the consistency of a second
 * independent vantage, so it never reaches a promotable claim.
 *
 * @param nodes      the ears observed for this candidate event
 * @param hypothesis the source position/emission-time being tested
 * @param quorumK    the number of mutually-consistent ears required to promote
 *                   (must be >= 2 to inherit the Lean single-ear refutation)
 * @param tol        the arrival-residual tolerance
 */
export function admitCoincidence(
  nodes: readonly MeshNode[],
  hypothesis: SourceHypothesis,
  quorumK: number,
  tol: number,
): CrossNodeCoincidenceAdmission {
  const consistentNodes: number[] = [];
  for (let i = 0; i < nodes.length; i++) {
    if (consistent(tol, hypothesis, nodes[i])) {
      consistentNodes.push(i);
    }
  }
  const coincidenceCount = consistentNodes.length;

  // A quorum below two cannot inherit the Lean single-ear refutation
  // (`single_ear_not_admitted` forbids admission at K = 1), so the gate
  // refuses to promote regardless of the count.
  const quorumIsSane = quorumK >= 2;
  const quorumMet = quorumIsSane && coincidenceCount >= quorumK;

  if (coincidenceCount === 0) {
    return {
      outcome: 'claim_withheld',
      consistentNodes,
      coincidenceCount,
      quorumMet: false,
      summary: 'no ear is consistent with the hypothesis — nothing to claim',
    };
  }

  if (quorumMet) {
    return {
      outcome: 'claim_admitted',
      consistentNodes,
      coincidenceCount,
      quorumMet: true,
      summary: `quorum K=${quorumK} met: ${coincidenceCount} independent ears mutually consistent with one source`,
    };
  }

  return {
    outcome: 'observation_only',
    consistentNodes,
    coincidenceCount,
    quorumMet: false,
    summary: quorumIsSane
      ? `unquorate detection: ${coincidenceCount} consistent ear(s) below quorum K=${quorumK}`
      : `quorum K=${quorumK} < 2 cannot self-certify a single ear`,
  };
}

/* ------------------------------------------------------------------ */
/*  Concrete corridor fixture (mirrors the Lean fixtures)             */
/* ------------------------------------------------------------------ */

/** Source at position 4, emitting at tick 100. Mirrors `CrossNodeCoincidence.hyp`. */
export const hyp: SourceHypothesis = { position: 4, emissionTime: 100 };

/** Ear at the near end: true arrival `100 + |4-0| = 104`. Mirrors `nodeA`. */
export const nodeA: MeshNode = {
  position: 0,
  reportedArrival: 104,
  detected: true,
};

/** Ear at the source: true arrival `100 + 0 = 100`. Mirrors `nodeB`. */
export const nodeB: MeshNode = {
  position: 4,
  reportedArrival: 100,
  detected: true,
};

/** Ear at the far end: true arrival `100 + |4-10| = 106`. Mirrors `nodeC`. */
export const nodeC: MeshNode = {
  position: 10,
  reportedArrival: 106,
  detected: true,
};

/**
 * A spoofer at position 20 claiming it heard the event at tick 100 — but the
 * geometry the real ears confirm predicts arrival `100 + 16 = 116` there, so
 * its residual is 16 and it is rejected at any sane tolerance.
 * Mirrors `CrossNodeCoincidence.spoofer`.
 */
export const spoofer: MeshNode = {
  position: 20,
  reportedArrival: 100,
  detected: true,
};

/** The three honest ears. Mirrors `CrossNodeCoincidence.realNodes`. */
export const realNodes: readonly MeshNode[] = [nodeA, nodeB, nodeC];

/**
 * Cross-Node Coincidence master witness. Bundles the boundary facts the Lean
 * theorem `mesh_coincidence_master` proves, each recomputed here from first
 * principles (never imported from Lean IR).
 *
 * Mirrors `CrossNodeCoincidence.mesh_coincidence_master`.
 */
export const CrossNodeCoincidenceMaster = {
  /** The real event is admitted at quorum two. */
  realEventAdmittedAtTwo:
    admitCoincidence(realNodes, hyp, 2, 1).outcome === 'claim_admitted',

  /** The real event reaches full quorum (all three ears agree). */
  realEventAdmittedAtThree:
    admitCoincidence(realNodes, hyp, 3, 1).outcome === 'claim_admitted',

  /** A single honest ear, alone, is not admitted at quorum two. */
  loneEarNotAdmitted:
    admitCoincidence([nodeA], hyp, 2, 1).outcome !== 'claim_admitted',

  /** The spoofer is inconsistent with the hypothesis the real ears confirm. */
  spooferRejected: consistent(1, hyp, spoofer) === false,

  /** The spoofer cannot inflate the coincidence: count over real ears plus the
   *  spoofer equals the count over the real ears alone. */
  spooferDoesNotInflate:
    countConsistent(1, hyp, [...realNodes, spoofer]) ===
    countConsistent(1, hyp, realNodes),

  /** Distinct vantages of the fixture source give distinct arrivals (104 != 106). */
  fixtureVantagesDistinct:
    predictedArrival(hyp, 0) !== predictedArrival(hyp, 10),
} as const;
