// @a0n/gnosis/distributed-inference-host/src/cross-modal-moonshine.ts
// File header (10-12 lines, MANDATORY): cross-modal Moonshine sieve.
// Audio and video are two images of the same source — their +1 Bule
// heartbeats should align frame-by-frame. The sieve scores per-frame
// agreement and emits a per-frame sigma adjustment hint for the video
// deblur pipeline. Heartbeat-aligned frames keep base sigma; disagreement
// frames (loud audio + still pixels, or quiet audio + flashing pixels)
// get sigma boosted to recover suspected motion blur. Mirrors the McKay-
// Moonshine framing where two distinct mathematical objects share a
// hidden symmetry — here the symmetry is the Bule heartbeat itself.
// Honest scope: this is a structural sieve, not a learned cross-modal
// model. Useful as a per-frame prior; not a substitute for proper
// audio-visual sync analysis.

export interface CrossModalSequenceInput {
  /** Per-frame visual scalar (e.g., mean luminance). */
  visual: Float32Array;
  /** Per-frame audio scalar (e.g., RMS over the matching audio window). */
  audio: Float32Array;
  /** They MUST have the same length — caller pre-aligns. */
}

export interface CrossModalScoreArgs extends CrossModalSequenceInput {
  /** Base sigma (the σ that would be used absent cross-modal info). */
  baseSigma: number;
  /** Adjustment strength. 0 = ignore cross-modal (always returns baseSigma).
   *  1 = max boost (sigma can up to double on full disagreement). Default 0.5. */
  adjustmentStrength?: number;
}

export interface CrossModalReport {
  /** Per-frame sigma recommendation. Length = visual.length. First frame
   *  is always baseSigma (no delta possible from index 0). */
  sigmaPerFrame: Float32Array;
  /** Per-frame agreement score in [0, 1]. 1 = perfectly aligned heartbeats. */
  agreementPerFrame: Float32Array;
  /** Aggregate Bule heartbeat counts. */
  heartbeat: {
    waste: number;        // frames where visual changed without audio (lighting transient)
    opportunity: number;  // frames where audio changed without visual (motion blur)
    diversity: number;    // frames where both changed similarly (true heartbeat)
  };
  /** Mean agreement score across all frames. */
  meanAgreement: number;
}

const HEARTBEAT_LOUD = 0.5;
const HEARTBEAT_STILL = 0.1;

/**
 * Cross-modal sigma scorer. Given paired per-frame visual + audio scalars
 * (already time-aligned by the caller), compute per-frame agreement of
 * their +1 Bule heartbeats and return a per-frame sigma hint suitable
 * for the deblur pipeline.
 */
export function crossModalSigmaScore(args: CrossModalScoreArgs): CrossModalReport {
  const { visual, audio, baseSigma } = args;
  const adjustmentStrength = args.adjustmentStrength ?? 0.5;

  if (visual.length !== audio.length) {
    throw new RangeError(
      `crossModalSigmaScore: visual.length (${visual.length}) !== audio.length (${audio.length})`,
    );
  }
  if (visual.length < 2) {
    throw new RangeError(
      `crossModalSigmaScore: need at least 2 frames, got ${visual.length}`,
    );
  }
  if (!(baseSigma > 0)) {
    throw new RangeError(
      `crossModalSigmaScore: baseSigma must be > 0, got ${baseSigma}`,
    );
  }
  if (!(adjustmentStrength >= 0 && adjustmentStrength <= 1)) {
    throw new RangeError(
      `crossModalSigmaScore: adjustmentStrength must be in [0, 1], got ${adjustmentStrength}`,
    );
  }

  const n = visual.length;

  // Per-frame absolute deltas. Index 0 is defined as 0 (no prior frame).
  const vDelta = new Float32Array(n);
  const aDelta = new Float32Array(n);
  let vMax = 0;
  let aMax = 0;
  for (let t = 1; t < n; t++) {
    const dv = Math.abs(visual[t] - visual[t - 1]);
    const da = Math.abs(audio[t] - audio[t - 1]);
    vDelta[t] = dv;
    aDelta[t] = da;
    if (dv > vMax) vMax = dv;
    if (da > aMax) aMax = da;
  }

  // Normalize each delta sequence to [0, 1]. Clamp 1 / 0 to 0 (no signal).
  const vNorm = vMax > 0 ? 1 / vMax : 0;
  const aNorm = aMax > 0 ? 1 / aMax : 0;

  const sigmaPerFrame = new Float32Array(n);
  const agreementPerFrame = new Float32Array(n);

  // Frame 0: no delta available, always baseSigma. Agreement defined as 0
  // (excluded from meanAgreement). Heartbeat: counted as diversity since
  // both deltas are zero — neither side beat.
  sigmaPerFrame[0] = baseSigma;
  agreementPerFrame[0] = 0;

  let waste = 0;
  let opportunity = 0;
  let diversity = 1; // frame 0
  let agreementSum = 0;

  for (let t = 1; t < n; t++) {
    const dV = vDelta[t] * vNorm;
    const dA = aDelta[t] * aNorm;
    const agreement = 1 - Math.abs(dV - dA);
    agreementPerFrame[t] = agreement;
    sigmaPerFrame[t] = baseSigma * (1 + adjustmentStrength * (1 - agreement));
    agreementSum += agreement;

    if (dV > HEARTBEAT_LOUD && dA < HEARTBEAT_STILL) {
      waste++;
    } else if (dA > HEARTBEAT_LOUD && dV < HEARTBEAT_STILL) {
      opportunity++;
    } else {
      diversity++;
    }
  }

  const meanAgreement = agreementSum / (n - 1);

  return {
    sigmaPerFrame,
    agreementPerFrame,
    heartbeat: { waste, opportunity, diversity },
    meanAgreement,
  };
}
