// @a0n/gnosis/distributed-inference-host/src/deblur-any-image.ts
// @ts-nocheck
// Public API for image-carrier deblur over arbitrary inputs.
//
// Pipeline:
//   1. Decode (file path / Buffer / pre-decoded plane) → linear Float32 in [0,1].
//   2. Build a luminance plane from the input (used only for σ selection so the
//      three colour channels share one principled σ rather than drifting apart).
//   3. Mirror-pad the luminance plane to next power-of-2 dims.
//   4. If `sigma === 'auto'` (default): selectSigmaByGoldenAttractor over a
//      φ-spaced grid; otherwise use the caller-supplied σ.
//   5. Mirror-pad each colour channel, run wienerDeblur with the chosen σ,
//      crop back to original dims.
//   6. Recompose channels and return.
//
// Notes:
//  - PSF is shared across channels; deblurring R/G/B independently with
//    different σ would amplify per-channel noise asymmetrically.
//  - Output is NOT clamped to [0, 1]. Wiener regularization can push values
//    slightly outside that range. `encodeRecoveredAsPNG` clamps for display
//    only, leaving the in-memory plane honest.
//  - The σ-selection σ-attractor (φ deficit-ratio fold) lives in
//    `sigma-attractor.ts`; it imports `interfere()` from
//    `@a0n/aether/src/glossolalia-moa.ts` via the wrapper in `attractor.ts`.
//  - API surface for `bettiSieve` and `rejectionWeights` is reserved;
//    meaningful only when called through `forkRaceFoldDeblur` (no-op + warn
//    on this code path).

import {
  ImageCarrierPipeline,
  type ImageCarrierAnalysis,
} from './image-carrier-pipeline';
import { gaussianPSF, recommendPsfSize } from './image-psf';
import { PHI } from './attractor';
import { wienerDeblur, computeDeblurCertificate, meshResidualDeblur, type DeblurCertificate, type MeshResidualDeblurConfig } from './image-deblur';
import { richardsonLucyDeblur } from './image-deblur-rl';
import { bilateralFilter } from './image-bilateral';
import { estimateNoiseFloor } from './image-noise-estimate';
import { mirrorPad, cropCenter } from './image-pad';
import {
  topologyAlphaStyleBackendFromImagingWasm,
  tryWasmMeshDepthConfidence,
  type StyleTransferImagingWasmSurface,
} from './style-transfer-imaging-wasm';
import {
  selectSigmaByGoldenAttractor,
  type SelectSigmaResult,
} from './sigma-attractor';
import { classifyDeblurStep, type DeblurStepClassification } from './image-spectrum-classify';
import { composeWithFaceUnit } from './cost-algebra';
import {
  TOWER_CEILINGS,
  vacuumBuleyProjection,
  type BuleyFace,
  type BuleyProjection,
} from './spectral-noise-monitor';
import {
  applyInvisibleWatermarkToPlane,
  floatPlaneToUint8Clamped,
  type InvisibleWatermarkOptions,
  type InvisibleWatermarkProvenance,
} from './invisible-watermark';

export type DeblurAnyImageInput =
  | string
  | Buffer
  | Uint8Array
  | { data: Float32Array | Uint8Array; width: number; height: number; channels: 1 | 3 };

export interface DeblurAnyImageOptions {
  /** Explicit σ in pixels, or `'auto'` to use the φ-attractor selector (default). */
  sigma?: number | 'auto';
  /** RGB recovers all three channels; luminance recovers Y only and rejoins
   *  the original chroma. Default 'rgb'. */
  channelMode?: 'rgb' | 'luminance';
  /** Deblur algorithm. 'wiener' is single-shot (fast); 'richardson-lucy' is
   *  iterative (slower, but less ringing on textured regions). 'mesh-residual' is topology-gated lift.
   *  Default 'wiener'. */
  method?: 'wiener' | 'richardson-lucy' | 'mesh-residual';
  /** Wiener regularization in variance units, or `'auto'` to estimate via
   *  Immerkær's Laplacian-of-Laplacian operator. Default 1e-3. Ignored when
   *  method !== 'wiener'. */
  noiseFloor?: number | 'auto';
  /** Richardson-Lucy iteration count. Default 20. Ignored when method === 'wiener'. */
  iterations?: number;
  /** PSF kernel side length (positive odd integer). When omitted, sized from σ
   *  via `recommendPsfSize` to keep truncation tails below the noise floor. */
  psfSize?: number;
  /** Edge-preserving post-pass on the recovered Y plane (suppresses residual
   *  Wiener / RL ringing in flat regions without softening real edges).
   *  `true` enables with defaults; pass `{sigmaSpatial, sigmaIntensity}` to
   *  tune; `false` (default) skips. */
  bilateralPostpass?:
    | boolean
    | { sigmaSpatial?: number; sigmaIntensity?: number; kernelRadius?: number };
  /** Chroma-ratio clamp in luminance mode. Bounds the per-pixel
   *  `recoveredY/originalY` ratio to keep colours from blowing out near
   *  near-black pixels. Default `[0.5, 2.0]`. Pass `false` to disable. */
  chromaRatioClamp?: [number, number] | false;
  /** Tower ceiling on the per-channel `carrierBuley.score`. Default `TOWER_CEILINGS.hexon` (6). */
  ceiling?: number;
  /** Apply Betti-Number Signature Sieve as a topology filter on the
   *  recovered plane. Currently no-op for single-config deblurAnyImage —
   *  meaningful when used through forkRaceFoldDeblur (the ensemble path).
   *  Reserved API surface; warns once if set on this code path. */
  bettiSieve?: boolean | { ceiling?: number; thresholdFraction?: number };
  /** Apply rejection-aware Buleyean weighting in the fold. Currently
   *  no-op for single-config deblurAnyImage — meaningful in
   *  forkRaceFoldDeblur. Reserved API surface; warns once if set. */
  rejectionWeights?: boolean | { alpha?: number };
  /** Optional invisible watermark embedding on the final recovered plane. */
  invisibleWatermark?: InvisibleWatermarkOptions;
  /** Optional frame index salt for deterministic per-frame video marks. */
  watermarkFrameIndex?: number;
  /** Required when method === 'mesh-residual' or meshGated === true to fetch topology masks. */
  wasmSurface?: StyleTransferImagingWasmSurface;
  /** Config for method === 'mesh-residual' or meshGated === true. */
  meshResidualConfig?: Partial<MeshResidualDeblurConfig>;
  /** If true, uses the topology gate to linearly blend the Wiener/RL result back with the original, preventing flat-field noise. */
  meshGated?: boolean;
}

export interface DeblurChannelReport {
  face: BuleyFace;
  reason: DeblurStepClassification['reason'];
  lowBandDelta: number;
  highBandDelta: number;
  totalDelta: number;
  carrierBuley: BuleyProjection;
}

export interface DeblurAnyImageResult {
  /** Length width * height * channels, row-major, channel-interleaved. NOT clamped. */
  recovered: Float32Array;
  width: number;
  height: number;
  channels: 1 | 3;
  selectedSigma: number;
  /** The actual noiseFloor used (estimated value if `noiseFloor: 'auto'`). */
  noiseFloorUsed: number;
  /** Algorithm used ('wiener', 'richardson-lucy', 'mesh-residual'). */
  methodUsed: 'wiener' | 'richardson-lucy' | 'mesh-residual';
  /** Whether the bilateral post-pass ran. */
  bilateralApplied: boolean;
  /** Whether the mesh gate was applied to the output. */
  meshGatedApplied?: boolean;
  /** Present when `sigma === 'auto'`. */
  attractor?: SelectSigmaResult;
  /** Per-channel deblur step report. Length === channels. */
  perChannel: ReadonlyArray<DeblurChannelReport>;
  /** Carrier witness from `ImageCarrierPipeline.analyze` over the luminance input. */
  analysis: ImageCarrierAnalysis;
  /** Populated when invisible watermark embedding is enabled. */
  invisibleWatermark?: InvisibleWatermarkProvenance;
  /** Residual deblur evidence matching the worker-WASM mesh pipeline. */
  certificate?: DeblurCertificate;
}

const DEFAULT_NOISE_FLOOR = 1e-3;
// Default attractor grid spans σ_min=0.5 across 7 φ-spaced candidates;
// the largest is σ_min · φ⁶ ≈ 8.97. The kernel must capture ~99.7% energy
// for the LARGEST σ in the grid so truncation ringing stays below the
// noise floor across the whole sweep.
const DEFAULT_AUTO_GRID_MAX_SIGMA = 0.5 * Math.pow(PHI, 6);

// Process-wide one-shot guards for reserved-but-no-op options.
// Set to true after the first warning so subsequent calls stay quiet.
let WARNED_BETTI_NOOP = false;
let WARNED_REJECTION_NOOP = false;

/**
 * Handles the gnosis deblur Any Image workflow.
 */
export async function deblurAnyImage(
  input: DeblurAnyImageInput,
  options: DeblurAnyImageOptions = {},
): Promise<DeblurAnyImageResult> {
  const noiseFloorOption: number | 'auto' = options.noiseFloor ?? DEFAULT_NOISE_FLOOR;
  const method: 'wiener' | 'richardson-lucy' | 'mesh-residual' = options.method ?? 'wiener';
  const iterations = options.iterations ?? 20;
  const bilateralOption = options.bilateralPostpass ?? false;
  const chromaClamp: [number, number] | false = options.chromaRatioClamp === undefined
    ? [0.5, 2.0]
    : options.chromaRatioClamp;
  const ceiling = options.ceiling ?? TOWER_CEILINGS.hexon;
  const channelMode = options.channelMode ?? 'rgb';
  const sigmaOption: number | 'auto' = options.sigma ?? 'auto';

  // Reserved-API soft warnings: these options are accepted at the public
  // surface but only meaningful through forkRaceFoldDeblur (ensemble path).
  // Warn once per process so callers see the no-op without spam.
  if (options.bettiSieve !== undefined && options.bettiSieve !== false && !WARNED_BETTI_NOOP) {
    // eslint-disable-next-line no-console
    console.warn(
      'deblurAnyImage: bettiSieve option set but is a no-op for single-config recovery. ' +
      'Use forkRaceFoldDeblur for ensemble-level Betti filtering.',
    );
    WARNED_BETTI_NOOP = true;
  }
  if (options.rejectionWeights !== undefined && options.rejectionWeights !== false && !WARNED_REJECTION_NOOP) {
    // eslint-disable-next-line no-console
    console.warn(
      'deblurAnyImage: rejectionWeights option set but is a no-op for single-config recovery. ' +
      'Use forkRaceFoldDeblur for ensemble-level rejection-proximity weighting.',
    );
    WARNED_REJECTION_NOOP = true;
  }

  // Auto-scale PSF kernel with σ so truncation tails stay below the noise
  // floor. For 'auto' σ-selection, we must size the kernel for the largest
  // candidate in the φ-grid (otherwise small σ candidates are fine but the
  // tail σ candidates ring). For explicit σ, size for that σ exactly.
  const psfSize = options.psfSize
    ?? recommendPsfSize(sigmaOption === 'auto' ? DEFAULT_AUTO_GRID_MAX_SIGMA : sigmaOption);

  const decoded = await decodeInput(input);
  const { data, width, height, channels } = decoded;

  if (channels !== 1 && channels !== 3) {
    throw new RangeError(`deblurAnyImage: channels must be 1 or 3, got ${channels}`);
  }

  // Effective channels processed by the deblur loop.
  // 'luminance' mode only deblurs Y; the other channels are bicubic-untouched.
  const effectiveChannels: 1 | 3 = channelMode === 'luminance' ? 1 : channels;

  // Luminance plane drives σ selection so all colour channels share one σ.
  const luminance = channels === 1
    ? new Float32Array(data)
    : buildLuminance(data, width, height, channels);

  // Mirror-pad luminance for σ selection (must be pow2 for fft2).
  const padLum = mirrorPad({ plane: luminance, width, height });

  // Resolve noiseFloor: 'auto' uses Immerkær's noise estimator on the
  // padded luminance (which is what σ-selection sees too).
  const noiseFloor = noiseFloorOption === 'auto'
    ? estimateNoiseFloor({
        plane: padLum.plane,
        width: padLum.paddedWidth,
        height: padLum.paddedHeight,
      })
    : noiseFloorOption;

  // Carrier witness on the luminance input (moved here to allow pre-recovery score to modulate mesh gating).
  const pipeline = new ImageCarrierPipeline();
  const analysis = pipeline.analyze({
    data: luminance,
    width,
    height,
    channels: 1,
    sampleRange: 'unit',
  });
  const preRecoveryScore = analysis.summaries[0]?.totalBuleCost ?? 0;
  const meanBuleCost = preRecoveryScore / (width * height);

  let selectedSigma: number;
  let attractor: SelectSigmaResult | undefined;
  if (sigmaOption === 'auto') {
    attractor = selectSigmaByGoldenAttractor({
      plane: padLum.plane,
      w: padLum.paddedWidth,
      h: padLum.paddedHeight,
      psfSize,
      noiseFloor,
    });
    selectedSigma = attractor.refinedSigma;
  } else {
    if (!Number.isFinite(sigmaOption) || sigmaOption <= 0) {
      throw new RangeError(`deblurAnyImage: sigma must be > 0 or 'auto', got ${sigmaOption}`);
    }
    selectedSigma = sigmaOption;
  }

  const psf = gaussianPSF(selectedSigma, psfSize);

  // Recover each channel that participates in the deblur.
  const recoveredChannels: Float32Array[] = [];
  const perChannel: DeblurChannelReport[] = [];
  let finalRecovered: Float32Array;
  let certificate: DeblurCertificate;

  if (method === 'mesh-residual') {
    if (!options.wasmSurface) {
      throw new Error("deblurAnyImage: wasmSurface is required for method 'mesh-residual'");
    }
    if (!options.wasmSurface.wasmMeshResidualDeblurRgbF32) {
      throw new Error("deblurAnyImage: wasmMeshResidualDeblurRgbF32 missing from wasmSurface");
    }

    let eta = options.meshResidualConfig?.overpotential ?? 0.85;
    let sat = options.meshResidualConfig?.saturation_limit ?? 0.18;
    let floor = options.meshResidualConfig?.admission_floor ?? 0.2;
    
    // The previous implementation used meanBuleCost which was defined elsewhere,
    // but the wasm signature directly takes overpotential and others.
    // If meanBuleCost is available, we could modulate it.
    // But since meanBuleCost is calculated outside (or inside the ensemble),
    // we'll just pass the provided options.

    let targetData: Float32Array;
    if (data instanceof Uint8Array) {
      targetData = new Float32Array(data.length);
      for (let i = 0; i < data.length; i++) targetData[i] = data[i]! / 255;
    } else {
      targetData = data;
    }

    const res = options.wasmSurface.wasmMeshResidualDeblurRgbF32(
      targetData,
      width,
      height,
      eta,
      floor,
      sat,
      options.meshResidualConfig?.depth_confidence_blend ?? 0.5
    );
    
    // Parse the JSON certificate returned by WASM.
    const parsedCert = JSON.parse(res.certificate_json) as DeblurCertificate;
    
    // Per-channel classification requires separating the channels.
    for (let c = 0; c < channels; c++) {
      const srcChan = extractChannel(targetData as Float32Array, width, height, channels, c);
      const dstChan = extractChannel(res.rgb, width, height, channels, c);
      const padSrc = mirrorPad({ plane: srcChan, width, height });
      const padDst = mirrorPad({ plane: dstChan, width, height });
      
      const classification = classifyDeblurStep({
        before: padSrc.plane,
        after: padDst.plane,
        w: padDst.paddedWidth,
        h: padDst.paddedHeight,
      });
      const carrierBuley = composeWithFaceUnit(vacuumBuleyProjection, classification.face);
      perChannel.push({
        face: classification.face,
        reason: classification.reason,
        lowBandDelta: classification.lowBandDelta,
        highBandDelta: classification.highBandDelta,
        totalDelta: classification.totalDelta,
        carrierBuley,
      });
    }

    finalRecovered = res.rgb;
    certificate = parsedCert;

  } else {
    let gateArray: Float32Array | null = null;
    if (options.meshGated && options.wasmSurface) {
       const backend = topologyAlphaStyleBackendFromImagingWasm(options.wasmSurface);
       // Build gate on the padded luminance so it aligns with the padded recovery planes
       const padWidth = padLum.paddedWidth;
       const padHeight = padLum.paddedHeight;
       
       // Luma is single channel. The wasm expects 1 or 3.
       const lumaReq = await backend.buildTopologyAlphaForStyle({ data: padLum.plane, width: padWidth, height: padHeight, channels: 1 });
       const dReq = await tryWasmMeshDepthConfidence(options.wasmSurface, {
         data: padLum.plane, width: padWidth, height: padHeight, channels: 1, monocularCueBlend: options.meshResidualConfig?.depth_confidence_blend ?? 0.5
       });
       
       if (lumaReq && dReq) {
         let eta = options.meshResidualConfig?.overpotential ?? 0.85;
         let sat = options.meshResidualConfig?.saturation_limit ?? 0.18;
         let floor = options.meshResidualConfig?.admission_floor ?? 0.2;
         
         // Adaptive Overpotential via Buleyean Rejection Weights
         if (meanBuleCost > 1.2) {
           eta *= 0.5;
           sat = Math.min(sat, 0.05);
           floor = Math.max(0, floor - 0.1);
         }
         
         const mConfig = {
           overpotential: eta,
           admission_floor: floor,
           saturation_limit: sat,
           depth_confidence_blend: options.meshResidualConfig?.depth_confidence_blend ?? 0.5,
         };
         gateArray = new Float32Array(padWidth * padHeight);
         for (let i = 0; i < gateArray.length; i++) {
           const topo = Math.max(0, Math.min(1, lumaReq.alpha[i]!));
           const depth = Math.max(0, Math.min(1, dReq.confidence[i]!));
           const rawGate = topo * (1.0 - mConfig.depth_confidence_blend) + depth * mConfig.depth_confidence_blend;
           let gate = 0;
           if (rawGate > mConfig.admission_floor) {
             gate = Math.max(0, Math.min(1, (rawGate - mConfig.admission_floor) / Math.max(1e-6, 1.0 - mConfig.admission_floor)));
           }
           gateArray[i] = gate;
         }
       }
    }

    for (let c = 0; c < effectiveChannels; c++) {
      const sourcePlane = effectiveChannels === channels
        ? extractChannel(data, width, height, channels, c)
        : luminance;

      const pad = mirrorPad({ plane: sourcePlane, width, height });
      const recoveredPad = method === 'richardson-lucy'
        ? richardsonLucyDeblur({
            plane: pad.plane,
            w: pad.paddedWidth,
            h: pad.paddedHeight,
            psf,
            psfW: psfSize,
            psfH: psfSize,
            iterations,
          })
        : wienerDeblur({
            plane: pad.plane,
            w: pad.paddedWidth,
            h: pad.paddedHeight,
            psf,
            psfW: psfSize,
            psfH: psfSize,
            noiseFloor,
          });

      let postProcessed = recoveredPad;
      if (bilateralOption) {
        const bilatOpts = typeof bilateralOption === 'object' ? bilateralOption : {};
        postProcessed = bilateralFilter({
          plane: recoveredPad,
          width: pad.paddedWidth,
          height: pad.paddedHeight,
          sigmaSpatial: bilatOpts.sigmaSpatial,
          sigmaIntensity: bilatOpts.sigmaIntensity,
          kernelRadius: bilatOpts.kernelRadius,
        });
      }

      if (gateArray) {
        for (let i = 0; i < postProcessed.length; i++) {
           const gate = gateArray[i]!;
           postProcessed[i] = pad.plane[i]! * (1 - gate) + postProcessed[i]! * gate;
        }
      }

      const recovered = cropCenter({
        plane: postProcessed,
        paddedWidth: pad.paddedWidth,
        paddedHeight: pad.paddedHeight,
        originX: pad.originX,
        originY: pad.originY,
        cropWidth: width,
        cropHeight: height,
      });
      recoveredChannels.push(recovered);

      const classification = classifyDeblurStep({
        before: pad.plane,
        after: postProcessed,
        w: pad.paddedWidth,
        h: pad.paddedHeight,
      });
      const carrierBuley = composeWithFaceUnit(vacuumBuleyProjection, classification.face);
      if (carrierBuley.score > ceiling) {
        throw new Error(
          `deblurAnyImage: channel ${c} carrierBuley.score=${carrierBuley.score} exceeds ceiling=${ceiling} (face=${classification.face}). Tower contract violated.`,
        );
      }
      perChannel.push({
        face: classification.face,
        reason: classification.reason,
        lowBandDelta: classification.lowBandDelta,
        highBandDelta: classification.highBandDelta,
        totalDelta: classification.totalDelta,
        carrierBuley,
      });
    }

    if (effectiveChannels === channels) {
      finalRecovered = recomposeInterleaved(recoveredChannels, width, height, channels);
    } else if (channelMode === 'luminance' && channels === 3) {
      finalRecovered = recomposeFromLuminanceRatio({
        originalData: data,
        originalLuminance: luminance,
        recoveredLuminance: recoveredChannels[0]!,
        width,
        height,
        ratioClamp: chromaClamp,
      });
    } else {
      finalRecovered = recoveredChannels[0]!;
    }
    
    certificate = computeDeblurCertificate(data, finalRecovered, width, height, channels);
  }

  const watermarked = applyInvisibleWatermarkToPlane({
    plane: finalRecovered,
    width,
    height,
    channels,
    options: options.invisibleWatermark,
    frameIndex: options.watermarkFrameIndex,
  });

  return {
    recovered: watermarked.plane,
    width,
    height,
    channels,
    selectedSigma,
    noiseFloorUsed: noiseFloor,
    methodUsed: method,
    bilateralApplied: Boolean(bilateralOption),
    meshGatedApplied: Boolean(options.meshGated),
    attractor,
    perChannel,
    analysis,
    invisibleWatermark: watermarked.provenance,
    certificate,
  };
}

/** PNG-encode the recovered plane. Clamps to [0, 1] for display only — the
 *  in-memory `result.recovered` stays unclamped. */
export async function encodeRecoveredAsPNG(result: DeblurAnyImageResult): Promise<Buffer> {
  const sharp = (await import('sharp')).default;
  const bytes = floatPlaneToUint8Clamped(result.recovered);
  const buf = await sharp(bytes, {
    raw: { width: result.width, height: result.height, channels: result.channels },
  })
    .png()
    .toBuffer();
  return buf;
}

// ────────────────────────── internals ──────────────────────────

interface DecodedImage {
  data: Float32Array;
  width: number;
  height: number;
  channels: 1 | 3;
}

async function decodeInput(input: DeblurAnyImageInput): Promise<DecodedImage> {
  if (
    typeof input === 'object' &&
    input !== null &&
    !Buffer.isBuffer(input) &&
    !(input instanceof Uint8Array) &&
    'data' in input
  ) {
    const { data, width, height, channels } = input;
    if (channels !== 1 && channels !== 3) {
      throw new RangeError(`decodeInput: channels must be 1 or 3, got ${channels}`);
    }
    const expected = width * height * channels;
    if (data.length !== expected) {
      throw new RangeError(
        `decodeInput: data.length=${data.length} != width*height*channels=${expected}`,
      );
    }
    if (data instanceof Float32Array) {
      return { data: new Float32Array(data), width, height, channels };
    }
    const out = new Float32Array(expected);
    for (let i = 0; i < expected; i++) out[i] = data[i]! / 255;
    return { data: out, width, height, channels };
  }

  const sharp = (await import('sharp')).default;
  const sharpInput: string | Buffer = typeof input === 'string'
    ? input
    : Buffer.isBuffer(input) ? input : Buffer.from(input);
  const { data, info } = await sharp(sharpInput)
    .removeAlpha()
    .toColourspace('srgb')
    .raw()
    .toBuffer({ resolveWithObject: true });
  const channels = info.channels === 1 ? 1 : info.channels === 3 ? 3 : null;
  if (channels === null) {
    throw new RangeError(
      `decodeInput: sharp returned ${info.channels} channels; only 1 or 3 are supported`,
    );
  }
  const expected = info.width * info.height * channels;
  if (data.length !== expected) {
    throw new RangeError(
      `decodeInput: sharp raw length=${data.length} != width*height*channels=${expected}`,
    );
  }
  const out = new Float32Array(expected);
  for (let i = 0; i < expected; i++) out[i] = data[i]! / 255;
  return { data: out, width: info.width, height: info.height, channels };
}

function buildLuminance(
  data: Float32Array,
  width: number,
  height: number,
  channels: number,
): Float32Array {
  const n = width * height;
  const out = new Float32Array(n);
  for (let i = 0; i < n; i++) {
    const r = data[i * channels]!;
    const g = data[i * channels + 1]!;
    const b = data[i * channels + 2]!;
    out[i] = 0.2126 * r + 0.7152 * g + 0.0722 * b;
  }
  return out;
}

function extractChannel(
  data: Float32Array,
  width: number,
  height: number,
  channels: number,
  channel: number,
): Float32Array {
  const n = width * height;
  const out = new Float32Array(n);
  for (let i = 0; i < n; i++) out[i] = data[i * channels + channel]!;
  return out;
}

function recomposeInterleaved(
  channels: Float32Array[],
  width: number,
  height: number,
  channelCount: number,
): Float32Array {
  const n = width * height;
  const out = new Float32Array(n * channelCount);
  for (let i = 0; i < n; i++) {
    for (let c = 0; c < channelCount; c++) {
      out[i * channelCount + c] = channels[c]![i]!;
    }
  }
  return out;
}

function recomposeFromLuminanceRatio(args: {
  originalData: Float32Array;
  originalLuminance: Float32Array;
  recoveredLuminance: Float32Array;
  width: number;
  height: number;
  ratioClamp: [number, number] | false;
}): Float32Array {
  const { originalData, originalLuminance, recoveredLuminance, width, height, ratioClamp } = args;
  const n = width * height;
  const out = new Float32Array(n * 3);
  const [lo, hi] = ratioClamp || [-Infinity, Infinity];
  for (let i = 0; i < n; i++) {
    const yOrig = originalLuminance[i]!;
    const yRec = recoveredLuminance[i]!;
    // Avoid division by zero on near-black pixels — fall back to delta scaling.
    let ratio = yOrig > 1e-6 ? yRec / yOrig : 1.0;
    // Clamp prevents dark-pixel ratio explosions from blowing chroma out.
    if (ratioClamp) {
      if (ratio < lo) ratio = lo;
      else if (ratio > hi) ratio = hi;
    }
    out[i * 3] = originalData[i * 3]! * ratio;
    out[i * 3 + 1] = originalData[i * 3 + 1]! * ratio;
    out[i * 3 + 2] = originalData[i * 3 + 2]! * ratio;
  }
  return out;
}
