// @a0n/gnosis/distributed-inference-host/src/decode-quality-reporter.ts
/**
 * decode-quality-reporter.ts — the in-mesh DECODE QUALITY SELF-REPORT.
 *
 * Implements MESH_QUALITY_REPORTING_FRAMEWORK.md §2(b): a rolling-window
 * decode-quality grader that the TS supervisor feeds with per-token decode
 * timing. It computes sustained tok/s → a letter grade (AAA..F) against the
 * single source of truth `mesh-decode-budget.json`, ESTABLISHES the warm
 * steady-state grade per loaded model, and surfaces grade-DIP / grade-UP
 * transitions across band floors. Correctness gates everything: any
 * `divergentTokens > 0` ⇒ grade F regardless of speed.
 *
 * This module is PURE (no OTel/IO dependency) so the grading/dip/up logic is
 * fully unit-testable. The OTLP emit lives in `decode-quality-otel.ts`, and the
 * flag-gated wiring into the decode path lives in `bin/openai-server.ts`. With
 * the flag OFF the reporter is never constructed → byte-identical behavior.
 *
 * The grades, band floors, and per-component byte tables are LOADED from
 * `mesh-decode-budget.json` (data, not code) — re-grading after a
 * hardware/quant change is a JSON edit.
 */

import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';

/** The decode component breakdown (matches admit-decode-profile.py `comps`). */
export type DecodeComponent =
  | 'qkv_matvec'
  | 'o_matvec'
  | 'ffn_matvec'
  | 'logit_matvec';

export const DECODE_COMPONENTS: readonly DecodeComponent[] = [
  'qkv_matvec',
  'o_matvec',
  'ffn_matvec',
  'logit_matvec',
];

/** Quantization dtype the loaded weights read per token (selects byte table). */
export type DecodeDtype = 'q8' | 'f16' | 'q4k';

/** Letter grade. Best → worst; F = broken / divergent. */
export type DecodeGrade = 'AAA' | 'AA' | 'A' | 'B' | 'C' | 'D' | 'F';

/** Best → worst ordering. F is handled specially (0 tps or divergence). */
export const GRADE_ORDER: readonly DecodeGrade[] = [
  'AAA',
  'AA',
  'A',
  'B',
  'C',
  'D',
  'F',
];

/**
 * grade_ordinal for the OTel gauge (framework §3): AAA=6,AA=5,A=4,B=3,C=2,D=1,F=0.
 * Higher = better, so a forge tile can map ordinal → color monotonically.
 */
export const GRADE_ORDINAL: Readonly<Record<DecodeGrade, number>> = {
  AAA: 6,
  AA: 5,
  A: 4,
  B: 3,
  C: 2,
  D: 1,
  F: 0,
};

interface BudgetGradeBand {
  min_tps: number;
  label: string;
}

interface MeshDecodeBudget {
  model: string;
  grades: Record<DecodeGrade, BudgetGradeBand>;
  components_bytes_gb: Record<string, Partial<Record<DecodeDtype, number>>>;
}

/**
 * Load `mesh-decode-budget.json` — the single source of truth — at module init.
 * The file lives at the package root (one dir above `src/`). We resolve it
 * relative to this module (`import.meta.url`) and fall back to a small set of
 * candidate locations so the loader survives both the native ESM runtime and
 * the monster/gnode bundler (which relocates the compiled module). A static
 * `import ... assert {type:'json'}` is intentionally avoided because the
 * bundler does not relocate the JSON alongside the bundle.
 */
function loadBudget(): MeshDecodeBudget {
  const here = dirname(fileURLToPath(import.meta.url));
  const candidates = [
    join(here, '..', 'mesh-decode-budget.json'),
    join(here, 'mesh-decode-budget.json'),
    join(process.cwd(), 'mesh-decode-budget.json'),
    join(process.cwd(), 'distributed-inference-host', 'mesh-decode-budget.json'),
  ];
  for (const path of candidates) {
    try {
      return JSON.parse(readFileSync(path, 'utf8')) as MeshDecodeBudget;
    } catch {
      /* try next candidate */
    }
  }
  throw new Error(
    'decode-quality-reporter: could not load mesh-decode-budget.json from any candidate path',
  );
}

const BUDGET: MeshDecodeBudget = loadBudget();

/** The budget's model id (e.g. "mistral-7b"). */
export const BUDGET_MODEL: string = BUDGET.model;

/** Human-readable label for a grade, straight from the budget JSON. */
export function gradeLabel(grade: DecodeGrade): string {
  return BUDGET.grades[grade]?.label ?? '';
}

/** The band floor (min sustained tok/s) that HOLDS a given grade. */
export function gradeFloorTps(grade: DecodeGrade): number {
  return BUDGET.grades[grade]?.min_tps ?? 0;
}

/**
 * Bytes (GB) the named component reads per token for a dtype, from the budget.
 * `undefined` when the component/dtype pair is not tabulated.
 */
export function componentBytesGb(
  component: DecodeComponent,
  dtype: DecodeDtype,
): number | undefined {
  return BUDGET.components_bytes_gb[component]?.[dtype];
}

/**
 * Grade a sustained tok/s + divergence count → letter grade.
 *
 * Mirrors admit-decode-profile.py `grade_for` EXACTLY (the post-hoc gate and
 * the in-mesh self-report must agree): any divergence ⇒ F; tps<=0 ⇒ F;
 * 0<tps<1 ⇒ D; otherwise the highest band whose `min_tps <= tps`.
 */
export function gradeForTps(tps: number, divergentTokens = 0): DecodeGrade {
  if (divergentTokens > 0) return 'F';
  if (!Number.isFinite(tps) || tps <= 0) return 'F';
  if (tps < gradeFloorTps('D') || tps < 1.0) {
    // The budget tabulates D.min_tps = 0 (>0,<1 band); guard with the literal
    // 1.0 boundary so the (0,1) interval always grades D.
    return 'D';
  }
  for (const grade of ['AAA', 'AA', 'A', 'B', 'C'] as const) {
    if (tps >= gradeFloorTps(grade)) return grade;
  }
  return 'D';
}

/** ordinal int for the OTel `mesh.decode.grade_ordinal` gauge. */
export function gradeOrdinal(grade: DecodeGrade): number {
  return GRADE_ORDINAL[grade];
}

/** True when `a` is a strictly better grade than `b` (AAA best, F worst). */
export function isBetterGrade(a: DecodeGrade, b: DecodeGrade): boolean {
  return GRADE_ORDER.indexOf(a) < GRADE_ORDER.indexOf(b);
}

/** A per-component ms/token + achieved GB/s, for the component gauges. */
export interface DecodeComponentSample {
  component: DecodeComponent;
  /** measured ms/token spent in this component (summed over layers). */
  ms: number;
  /** achieved GB/s = budget bytes(dtype) / (ms/1000); 0 when untabulated. */
  gbs: number;
}

/** A grade transition the reporter surfaces as an OTel event. */
export interface GradeTransition {
  kind: 'grade_dip' | 'grade_up';
  from: DecodeGrade;
  to: DecodeGrade;
  tps: number;
  /** For dips: the component that most plausibly caused the slowdown. */
  causeComponent?: DecodeComponent;
}

/** Immutable snapshot of the reporter, suitable for an OTel emit. */
export interface DecodeQualitySnapshot {
  model: string;
  tps: number;
  grade: DecodeGrade;
  gradeOrdinal: number;
  /** The warm steady-state grade established for this loaded model, if any. */
  establishedGrade: DecodeGrade | null;
  divergentTokens: number;
  windowSize: number;
  components: readonly DecodeComponentSample[];
  /** A transition crossed on the token that produced this snapshot, if any. */
  transition?: GradeTransition;
}

export interface DecodeQualityReporterOptions {
  /** Rolling window length in decode tokens (default 32, per framework §2b). */
  windowSize?: number;
  /**
   * Tokens that must be observed before the warm steady-state grade is
   * ESTABLISHED. Until then no dip/up fires (avoids cold-start false alarms).
   * Default = windowSize (one full window).
   */
  establishAfterTokens?: number;
  /** Model id for the snapshot/resource attr. Default = budget model. */
  model?: string;
  /** Dtype selecting the per-component byte table. Default 'q8' (live lane). */
  dtype?: DecodeDtype;
  /** Optional override of the component-bytes lookup (testing). */
  componentBytesGb?: (
    component: DecodeComponent,
    dtype: DecodeDtype,
  ) => number | undefined;
}

/**
 * A rolling decode-quality grader.
 *
 * Hot-path cost per token is one ring-buffer write + a running-sum update + a
 * grade compare — no allocation, no branching beyond the compare. `record()`
 * returns a snapshot ONLY when something worth emitting changed (a transition,
 * or the first established grade); otherwise it returns `null` so the caller
 * can cheaply skip the OTLP emit. Use `snapshot()` to force a full read.
 */
export class DecodeQualityReporter {
  private readonly windowSize: number;
  private readonly establishAfterTokens: number;
  private readonly model: string;
  private readonly dtype: DecodeDtype;
  private readonly bytesLookup: (
    component: DecodeComponent,
    dtype: DecodeDtype,
  ) => number | undefined;

  // Ring buffer of per-token forward-ms.
  private readonly ring: Float64Array;
  private ringCount = 0;
  private ringHead = 0;
  private ringSumMs = 0;

  private tokensObserved = 0;
  private divergentTokens = 0;
  private establishedGrade: DecodeGrade | null = null;
  private lastGrade: DecodeGrade | null = null;

  // Latest per-component ms/token, if the caller supplies a breakdown.
  private lastComponentMs: Partial<Record<DecodeComponent, number>> = {};

  constructor(options: DecodeQualityReporterOptions = {}) {
    const windowSize = Math.max(1, Math.floor(options.windowSize ?? 32));
    this.windowSize = windowSize;
    this.establishAfterTokens = Math.max(
      1,
      Math.floor(options.establishAfterTokens ?? windowSize),
    );
    this.model = options.model ?? BUDGET_MODEL;
    this.dtype = options.dtype ?? 'q8';
    this.bytesLookup = options.componentBytesGb ?? componentBytesGb;
    this.ring = new Float64Array(windowSize);
  }

  /** Sustained tok/s over the rolling window (0 when no samples). */
  get sustainedTps(): number {
    if (this.ringCount === 0 || this.ringSumMs <= 0) return 0;
    const meanMs = this.ringSumMs / this.ringCount;
    return meanMs > 0 ? 1000 / meanMs : 0;
  }

  /** Current grade for the rolling window (correctness-gated). */
  get grade(): DecodeGrade {
    return gradeForTps(this.sustainedTps, this.divergentTokens);
  }

  get established(): DecodeGrade | null {
    return this.establishedGrade;
  }

  /**
   * Record one decoded token's forward latency (ms) and optional per-component
   * breakdown. Returns a snapshot when a transition fires OR the warm grade is
   * first established; `null` otherwise (the common, zero-emit path).
   *
   * @param forwardMs   wall-clock ms attributed to producing this token.
   * @param opts.divergent  set true when this token diverged from the
   *   reference (drives the hard correctness gate → F).
   * @param opts.componentMs  optional ms/token split across decode components.
   */
  record(
    forwardMs: number,
    opts: {
      divergent?: boolean;
      componentMs?: Partial<Record<DecodeComponent, number>>;
    } = {},
  ): DecodeQualitySnapshot | null {
    const ms = Number.isFinite(forwardMs) && forwardMs > 0 ? forwardMs : 0;

    // Ring-buffer running-sum update (O(1), no allocation).
    if (this.ringCount < this.windowSize) {
      this.ring[this.ringHead] = ms;
      this.ringSumMs += ms;
      this.ringCount++;
    } else {
      this.ringSumMs += ms - this.ring[this.ringHead]!;
      this.ring[this.ringHead] = ms;
    }
    this.ringHead = (this.ringHead + 1) % this.windowSize;

    this.tokensObserved++;
    if (opts.divergent) this.divergentTokens++;
    if (opts.componentMs) {
      this.lastComponentMs = { ...opts.componentMs };
    }

    const grade = this.grade;
    let transition: GradeTransition | undefined;
    let emit = false;

    // Establish the warm steady-state grade once enough tokens have flowed.
    if (
      this.establishedGrade === null &&
      this.tokensObserved >= this.establishAfterTokens
    ) {
      this.establishedGrade = grade;
      emit = true; // first established grade is worth surfacing
    } else if (this.establishedGrade !== null && grade !== this.establishedGrade) {
      // A transition across the established band floor.
      if (isBetterGrade(grade, this.establishedGrade)) {
        transition = {
          kind: 'grade_up',
          from: this.establishedGrade,
          to: grade,
          tps: this.sustainedTps,
        };
      } else {
        transition = {
          kind: 'grade_dip',
          from: this.establishedGrade,
          to: grade,
          tps: this.sustainedTps,
          causeComponent: this.slowestComponent(),
        };
      }
      // Re-establish at the new grade so we report each band crossing once,
      // not on every subsequent token at the degraded/upgraded level.
      this.establishedGrade = grade;
      emit = true;
    }

    this.lastGrade = grade;
    return emit ? this.snapshot(transition) : null;
  }

  /**
   * The component most likely responsible for a slowdown: the one whose
   * measured ms/token most exceeds its budget (achieved GB/s farthest below the
   * dtype's expectation). Falls back to the largest absolute ms when no byte
   * table is available (FFN dominates by construction). Returns undefined when
   * no component breakdown has been recorded.
   */
  private slowestComponent(): DecodeComponent | undefined {
    const entries = Object.entries(this.lastComponentMs) as Array<
      [DecodeComponent, number]
    >;
    if (entries.length === 0) return undefined;
    let worst: DecodeComponent | undefined;
    let worstScore = -Infinity;
    for (const [component, msVal] of entries) {
      if (!(msVal > 0)) continue;
      const gb = this.bytesLookup(component, this.dtype);
      // Score = ms relative to its byte budget; a component reading few bytes
      // but burning lots of ms is the regression. Without bytes, score = ms.
      const score = gb && gb > 0 ? msVal / gb : msVal;
      if (score > worstScore) {
        worstScore = score;
        worst = component;
      }
    }
    return worst;
  }

  /** Per-component ms + achieved GB/s for the component gauges. */
  componentSamples(): DecodeComponentSample[] {
    const samples: DecodeComponentSample[] = [];
    for (const component of DECODE_COMPONENTS) {
      const ms = this.lastComponentMs[component];
      if (ms === undefined) continue;
      const gb = this.bytesLookup(component, this.dtype);
      const gbs = gb && ms > 0 ? gb / (ms / 1000) : 0;
      samples.push({ component, ms, gbs });
    }
    return samples;
  }

  /** Full read of the current state (forces an emit-shaped snapshot). */
  snapshot(transition?: GradeTransition): DecodeQualitySnapshot {
    const grade = this.grade;
    return {
      model: this.model,
      tps: this.sustainedTps,
      grade,
      gradeOrdinal: gradeOrdinal(grade),
      establishedGrade: this.establishedGrade,
      divergentTokens: this.divergentTokens,
      windowSize: this.windowSize,
      components: this.componentSamples(),
      ...(transition ? { transition } : {}),
    };
  }
}
