// @a0n/gnosis/distributed-inference-host/src/decode-quality-otel.ts
/**
 * decode-quality-otel.ts — flag-gated OTLP emit for the decode self-report.
 *
 * Bridges `DecodeQualitySnapshot` (pure, from decode-quality-reporter.ts) to
 * OpenTelemetry, REUSING the same OTLP backend `apps/aeon-forge/src/telemetry/
 * tracing.ts` exports (the `@a0n/telemetry` surface over
 * `OTEL_EXPORTER_OTLP_ENDPOINT`). aeon-forge already consumes that backend, so
 * the mesh emits onto the same pane.
 *
 * THE SCHEMA (MESH_QUALITY_REPORTING_FRAMEWORK.md §3):
 *   resource attrs : service.name=gnosis-openai-mesh, mesh.model, mesh.revision
 *   gauges         : mesh.decode.tps, mesh.decode.grade_ordinal (+attr grade)
 *                    mesh.decode.component.gbs{component=...}
 *                    mesh.decode.component.ms{component=...}
 *   hard attr      : mesh.decode.divergent_tokens (>0 ⇒ grade F)
 *   events         : mesh.decode.grade_dip{from,to,cause_component}
 *                    mesh.decode.grade_up{from,to}
 *
 * SUBSTITUTION NOTE: `@a0n/telemetry` exposes the TRACE (span) API only — there
 * is no Meter/metrics exporter in the repo's telemetry surface, and inside
 * gnosis the package is stubbed. So this emitter encodes the schema onto a
 * span: each emit opens a short `mesh.decode.report` span whose attributes carry
 * the gauge values under the §3 metric names, and dip/up land as span EVENTS
 * with the §3 event names. A forge consumer reads identical keys; if/when a
 * Meter lands in `@a0n/telemetry`, swap the span attrs for true gauges without
 * changing the names. Default OFF + best-effort: any telemetry failure (stubbed
 * package, no endpoint) degrades to a console log, never to a thrown error on
 * the decode path.
 */

import type {
  DecodeComponentSample,
  DecodeQualitySnapshot,
  GradeTransition,
} from './decode-quality-reporter';

export const MESH_DECODE_SERVICE_NAME = 'gnosis-openai-mesh';

/** Metric names — the contract between mesh → forge (framework §3). */
export const MESH_DECODE_METRICS = {
  tps: 'mesh.decode.tps',
  gradeOrdinal: 'mesh.decode.grade_ordinal',
  componentGbs: 'mesh.decode.component.gbs',
  componentMs: 'mesh.decode.component.ms',
  divergentTokens: 'mesh.decode.divergent_tokens',
  gradeDip: 'mesh.decode.grade_dip',
  gradeUp: 'mesh.decode.grade_up',
} as const;

export interface DecodeQualityOtelOptions {
  /** mesh.model resource attr (e.g. "mistral-7b"). */
  model: string;
  /** mesh.revision resource attr (deployed build/rev id). */
  revision: string;
  /** Override the OTLP endpoint resolution (defaults to env). */
  env?: Record<string, unknown>;
  /** Mark dev so the console span exporter also mirrors locally. */
  debug?: boolean;
}

/** Minimal structural view of `@a0n/telemetry` so this stays decoupled. */
interface TelemetrySurface {
  getTracer(name: string): TracerLike;
  resolveTraceExporterTarget(env: unknown): { endpoint: string };
  createNodeTracerProvider(opts: unknown): TraceProviderLike;
  registerTracerProvider(provider: TraceProviderLike): void;
  shutdownTracerProvider(provider: TraceProviderLike | null): Promise<void>;
}
interface SpanLike {
  setAttribute(key: string, value: string | number | boolean): void;
  addEvent(name: string, attrs?: Record<string, string | number | boolean>): void;
  end(): void;
}
interface TracerLike {
  startSpan(name: string): SpanLike;
}
type TraceProviderLike = object;

/**
 * The flag-gated emitter. Construct ONLY when MESH_QUALITY_REPORT=1; with the
 * flag off this is never instantiated, so the decode path is byte-identical and
 * pays zero telemetry cost. Initialization is lazy + best-effort: if the
 * telemetry package is stubbed/absent or no OTLP endpoint is configured, the
 * emitter falls back to a structured console log and never throws.
 */
export class DecodeQualityOtelEmitter {
  private readonly model: string;
  private readonly revision: string;
  private readonly env: Record<string, unknown>;
  private readonly debug: boolean;

  private telemetry: TelemetrySurface | null = null;
  private tracer: TracerLike | null = null;
  private provider: TraceProviderLike | null = null;
  private initialized = false;
  private telemetryUsable = false;

  constructor(options: DecodeQualityOtelOptions) {
    this.model = options.model;
    this.revision = options.revision;
    this.env = options.env ?? (typeof process !== 'undefined' ? process.env : {});
    this.debug = options.debug ?? false;
  }

  /**
   * Best-effort lazy init of the OTLP tracer provider. Returns false (and stays
   * false) when telemetry is unavailable — callers then log instead of emit.
   */
  private async ensureInit(): Promise<boolean> {
    if (this.initialized) return this.telemetryUsable;
    this.initialized = true;
    if (!hasExplicitTraceExporterTarget(this.env)) {
      return false;
    }
    try {
      // Dynamic import so a stubbed/absent `@a0n/telemetry` (gnosis ships a
      // no-op stub) never breaks the import graph of the decode path.
      const mod = (await import('@a0n/telemetry')) as unknown as TelemetrySurface;
      if (
        typeof mod?.getTracer !== 'function' ||
        typeof mod?.createNodeTracerProvider !== 'function' ||
        typeof mod?.resolveTraceExporterTarget !== 'function'
      ) {
        return false; // stub — degrade to console
      }
      const target = mod.resolveTraceExporterTarget(this.env);
      const provider = mod.createNodeTracerProvider({
        endpoint: target.endpoint,
        debug: this.debug,
        resourceAttributes: {
          'service.name': MESH_DECODE_SERVICE_NAME,
          'mesh.model': this.model,
          'mesh.revision': this.revision,
        },
      });
      mod.registerTracerProvider(provider);
      this.telemetry = mod;
      this.provider = provider;
      this.tracer = mod.getTracer('gnosis-openai-mesh-decode');
      this.telemetryUsable = true;
    } catch {
      this.telemetryUsable = false;
    }
    return this.telemetryUsable;
  }

  /** Emit one snapshot (gauges + any transition event). Never throws. */
  async emit(snapshot: DecodeQualitySnapshot): Promise<void> {
    try {
      const ok = await this.ensureInit();
      if (!ok || !this.tracer) {
        this.logFallback(snapshot);
        return;
      }
      const span = this.tracer.startSpan('mesh.decode.report');
      span.setAttribute('mesh.model', this.model);
      span.setAttribute('mesh.revision', this.revision);
      span.setAttribute(MESH_DECODE_METRICS.tps, round(snapshot.tps));
      span.setAttribute(MESH_DECODE_METRICS.gradeOrdinal, snapshot.gradeOrdinal);
      span.setAttribute('grade', snapshot.grade);
      span.setAttribute(
        MESH_DECODE_METRICS.divergentTokens,
        snapshot.divergentTokens,
      );
      if (snapshot.establishedGrade) {
        span.setAttribute('mesh.decode.established_grade', snapshot.establishedGrade);
      }
      for (const c of snapshot.components) {
        // component=… is encoded into the metric key suffix per §3's
        // `{component=...}` attribute-selector notation.
        span.setAttribute(componentKey(MESH_DECODE_METRICS.componentMs, c), round(c.ms));
        span.setAttribute(componentKey(MESH_DECODE_METRICS.componentGbs, c), round(c.gbs));
      }
      if (snapshot.transition) {
        this.emitTransitionEvent(span, snapshot.transition);
      }
      span.end();
    } catch {
      this.logFallback(snapshot);
    }
  }

  private emitTransitionEvent(span: SpanLike, t: GradeTransition): void {
    if (t.kind === 'grade_dip') {
      span.addEvent(MESH_DECODE_METRICS.gradeDip, {
        from: t.from,
        to: t.to,
        tps: round(t.tps),
        ...(t.causeComponent ? { cause_component: t.causeComponent } : {}),
      });
    } else {
      span.addEvent(MESH_DECODE_METRICS.gradeUp, {
        from: t.from,
        to: t.to,
        tps: round(t.tps),
      });
    }
  }

  /** Structured console fallback so the signal is never silently lost. */
  private logFallback(snapshot: DecodeQualitySnapshot): void {
    const t = snapshot.transition;
    const evt = t ? ` ${t.kind} ${t.from}->${t.to}${t.causeComponent ? ` cause=${t.causeComponent}` : ''}` : '';
    console.log(
      `[mesh.decode.report] model=${snapshot.model} rev=${this.revision} ` +
        `tps=${round(snapshot.tps)} grade=${snapshot.grade} ordinal=${snapshot.gradeOrdinal} ` +
        `divergent=${snapshot.divergentTokens}${evt}`,
    );
  }

  /** Flush + shutdown the provider (call on server teardown). Never throws. */
  async shutdown(): Promise<void> {
    try {
      if (this.telemetry && this.provider) {
        await withTimeout(
          this.telemetry.shutdownTracerProvider(this.provider),
          1000
        );
      }
    } catch {
      /* ignore */
    } finally {
      this.provider = null;
      this.tracer = null;
      this.telemetryUsable = false;
    }
  }
}

function componentKey(base: string, sample: DecodeComponentSample): string {
  return `${base}.${sample.component}`;
}

function round(n: number): number {
  return Math.round(n * 1000) / 1000;
}

function hasExplicitTraceExporterTarget(env: Record<string, unknown>): boolean {
  const keys = [
    'OTEL_EXPORTER_OTLP_TRACES_ENDPOINT',
    'OTEL_EXPORTER_OTLP_ENDPOINT',
    'DATADOG_OTLP_TRACES_ENDPOINT',
    'DATADOG_OTLP_ENDPOINT',
    'DD_OTLP_TRACES_ENDPOINT',
    'DD_OTLP_ENDPOINT',
    'DATADOG_API_KEY',
    'DD_API_KEY',
  ];
  return keys.some((key) => {
    const value = env[key];
    return typeof value === 'string' && value.trim().length > 0;
  });
}

async function withTimeout<T>(promise: Promise<T>, timeoutMs: number): Promise<T | void> {
  let timer: ReturnType<typeof setTimeout> | null = null;
  try {
    return await Promise.race([
      promise,
      new Promise<void>((resolve) => {
        timer = setTimeout(resolve, timeoutMs);
      }),
    ]);
  } finally {
    if (timer) clearTimeout(timer);
  }
}

/** True when the in-mesh self-report flag is on. Default OFF. */
export function meshQualityReportEnabled(
  env: Record<string, unknown> = typeof process !== 'undefined' ? process.env : {},
): boolean {
  const v = env['MESH_QUALITY_REPORT'];
  return v === '1' || v === 'true' || v === 'on';
}
