// @a0n/gnosis/distributed-inference-host/src/electrical-network-frequency.ts
/**
 * # Electrical Network Frequency — embedded timestamp, geolocation, tamper carrier
 *
 * Mirrors `Gnosis.Mesh.ElectricalNetworkFrequency` from `open-source/gnosis-math`,
 * the Rustic Church carrier for the ENF side channel: the mains hum of an
 * electrical grid drifts in tiny excursions around its nominal frequency, those
 * excursions are essentially identical across an entire grid region at any
 * instant, and any recording made near mains-powered equipment captures a faint
 * copy of that drift. The captured drift acts as a covert side channel:
 *
 *  - a recovered ENF trace, matched against a grid's logged reference, pins
 *    *when* the recording was made (timestamp) and *which* grid region it came
 *    from (geolocation to grid);
 *  - an authentic trace must actually *vary* — a flat, constant trace carries no
 *    drift fingerprint and cannot timestamp anything;
 *  - a spliced or edited recording breaks ENF continuity, leaving a
 *    discontinuity (a jump) at the cut that a continuous authentic trace never
 *    shows.
 *
 * ## What is (and is not) carried
 *
 * This module is the runtime promotion check for an ENF observation. It carries
 * ONLY the finite contract the Lean file proves:
 *
 *  - the saturating excursion metric (`adist`);
 *  - a pointwise length-and-tolerance grid match (`matches`);
 *  - the variation gate as the source-only-rejection analog: a constant trace
 *    has zero total variation and carries no timestamp
 *    (`totalVariation` / `hasCutoff` / `variationAdmits`);
 *  - the continuity / tamper gate: a splice leaves a successive-sample jump
 *    exceeding the step bound (`continuity` / `tamperFlagged`).
 *
 * Real STFT frequency estimation, clock alignment to the grid log, the true grid
 * topology, harmonic extraction, and noise robustness stay runtime/empirical
 * obligations OUTSIDE this certificate — they are not theorems and are not
 * asserted here.
 *
 * The promotion check folds the contract into the same epistemic-outcome
 * vocabulary the sibling mesh carriers already speak
 * (`claim_admitted` / `observation_only` / `claim_withheld`), so an ENF
 * observation cannot be displayed as a timestamp/geolocation claim until it has
 * cleared the Lean-side variation/match/continuity gate.
 *
 * ## Lean source-of-truth
 *
 * @see open-source/gnosis-math/Gnosis/Mesh/ElectricalNetworkFrequency.lean
 * @see open-source/gnosis-math/Gnosis/Mesh/CrossNodeCoincidence.lean (mesh sibling)
 */

/* ------------------------------------------------------------------ */
/*  Excursion distance and traces                                     */
/* ------------------------------------------------------------------ */

/**
 * An ENF trace is a list of per-tick frequency excursions (small centered
 * offsets recovered from a recording). A grid reference is the same shape: the
 * grid's logged excursion sequence over the same window. Excursions are
 * non-negative integers (saturating `Nat` arithmetic), matching the Lean
 * `Trace`.
 */
export type Trace = readonly number[];

/** Absolute difference on `Nat` via saturating subtraction (`|a - b|`). The
 *  per-tick excursion residual metric, shared with the mesh corridor carriers.
 *  Mirrors `ElectricalNetworkFrequency.adist`. */
export function adist(a: number, b: number): number {
  return Math.max(0, a - b) + Math.max(0, b - a);
}

/**
 * Pointwise match: the trace and grid have equal length and every aligned pair
 * of excursions is within tolerance. A length mismatch is an immediate
 * rejection. Mirrors `ElectricalNetworkFrequency.matchesB`.
 */
export function matches(tol: number, trace: Trace, grid: Trace): boolean {
  if (trace.length !== grid.length) {
    return false;
  }
  for (let i = 0; i < trace.length; i++) {
    if (adist(trace[i], grid[i]) > tol) {
      return false;
    }
  }
  return true;
}

/* ------------------------------------------------------------------ */
/*  Variation gate (source-only rejection analog)                     */
/* ------------------------------------------------------------------ */

/**
 * Total variation of a trace: the summed excursion distance between successive
 * samples. A flat trace has zero total variation; a drifting trace has positive
 * total variation. Mirrors `ElectricalNetworkFrequency.totalVariation`.
 */
export function totalVariation(trace: Trace): number {
  let tv = 0;
  for (let i = 1; i < trace.length; i++) {
    tv += adist(trace[i - 1], trace[i]);
  }
  return tv;
}

/**
 * The variation gate: a trace is admissible as a timestamp carrier only when it
 * actually drifts. A constant trace fails this gate (zero total variation, no
 * drift, no timestamp). Mirrors `ElectricalNetworkFrequency.variationAdmits`.
 */
export function variationAdmits(trace: Trace): boolean {
  return totalVariation(trace) > 0;
}

/* ------------------------------------------------------------------ */
/*  Continuity / tamper gate                                          */
/* ------------------------------------------------------------------ */

/**
 * A trace is continuous at step bound `maxStep` when every successive pair of
 * samples is within `maxStep`. Mirrors
 * `ElectricalNetworkFrequency.continuousB`.
 */
export function continuity(maxStep: number, trace: Trace): boolean {
  for (let i = 1; i < trace.length; i++) {
    if (adist(trace[i - 1], trace[i]) > maxStep) {
      return false;
    }
  }
  return true;
}

/**
 * The maximum successive jump within a window — the cutoff the continuity gate
 * tests against `maxStep`. A spliced trace has a jump at the cut that a
 * continuous authentic trace never shows; `hasCutoff` reports whether the trace
 * jumps by more than `maxStep` anywhere (the negation of `continuity`).
 */
export function hasCutoff(maxStep: number, trace: Trace): boolean {
  return !continuity(maxStep, trace);
}

/** A splice is flagged when continuity fails: some successive jump exceeds the
 *  step bound. Mirrors `ElectricalNetworkFrequency.tamperFlagged`. */
export function tamperFlagged(maxStep: number, trace: Trace): boolean {
  return !continuity(maxStep, trace);
}

/* ------------------------------------------------------------------ */
/*  Runtime promotion check (epistemic outcome)                       */
/* ------------------------------------------------------------------ */

/**
 * Epistemic outcome of an ENF observation. Identical vocabulary to the sibling
 * mesh carriers (`./cross-node-coincidence`, `./active-acoustic-pose`), kept as
 * a local literal union so this certificate has no import coupling:
 *
 *  - `claim_admitted`    — a varied trace matches its grid reference AND is
 *                          continuous, so the timestamp/geolocation may be
 *                          promoted to a certified claim.
 *  - `observation_only`  — a varied trace matches its grid reference but is
 *                          tamper-suspect (a splice broke continuity), so the
 *                          match is present yet cannot be certified; the
 *                          observation is reported, never promoted.
 *  - `claim_withheld`    — the trace is constant (zero variation, no drift
 *                          fingerprint) or does not match the grid, so there is
 *                          no timestamp to claim.
 */
export type ENFEpistemicOutcome =
  | 'claim_admitted'
  | 'observation_only'
  | 'claim_withheld';

/** The reason an ENF claim was held back from promotion. */
export type ENFLimitationClass =
  | 'no_drift_fingerprint'
  | 'grid_mismatch'
  | 'tamper_suspected'
  | 'none';

export interface ENFAdmission {
  readonly outcome: ENFEpistemicOutcome;
  readonly limitation: ENFLimitationClass;
  /** Whether the trace actually drifts (passed the variation gate). */
  readonly varied: boolean;
  /** Whether the trace matched the grid reference within tolerance. */
  readonly gridMatched: boolean;
  /** Whether the trace stayed continuous (no splice flagged). */
  readonly continuous: boolean;
  /** The measured total variation (the displayable drift magnitude). */
  readonly totalVariation: number;
  readonly summary: string;
}

export interface ENFAdmitOptions {
  /** The grid reference to match the trace against (timestamp + geolocation). */
  readonly grid: Trace;
  /** Tolerance for the pointwise grid match. */
  readonly matchTol: number;
  /** Step bound for the continuity / tamper gate. */
  readonly maxStep: number;
}

/**
 * The promotion check Monster Mesh runs before any UI surface displays an ENF
 * observation as a timestamp/geolocation claim. It folds the Lean contract into
 * an epistemic outcome:
 *
 *  - the variation gate decides whether there is any drift fingerprint at all —
 *    a constant trace carries no timestamp and is withheld (the source-only
 *    rejection analog, mirroring `constantTrace_rejected`);
 *  - the grid match decides which grid/when the trace belongs to — a mismatch
 *    is withheld (mirroring `matchedTrace_matches_grid` /
 *    `matchedTrace_not_other_grid`);
 *  - the continuity gate decides whether the matched, varied trace may be
 *    *certified* — a splice (tamper flag) downgrades it to observation-only
 *    rather than a promotable claim (mirroring `splicedTrace_flagged`).
 *
 * @param trace the recovered ENF trace under examination
 * @param opts  the grid reference, match tolerance, and continuity step bound
 */
export function admitENF(trace: Trace, opts: ENFAdmitOptions): ENFAdmission {
  const tv = totalVariation(trace);
  const varied = tv > 0;
  const gridMatched = matches(opts.matchTol, trace, opts.grid);
  const continuous = continuity(opts.maxStep, trace);

  // Variation gate: a constant trace carries no drift fingerprint, so there is
  // no timestamp to claim regardless of how it lines up against the grid.
  if (!varied) {
    return {
      outcome: 'claim_withheld',
      limitation: 'no_drift_fingerprint',
      varied,
      gridMatched,
      continuous,
      totalVariation: tv,
      summary:
        'constant trace: zero total variation carries no drift fingerprint — nothing to timestamp',
    };
  }

  // Match gate: a varied trace that does not line up with the grid reference
  // pins no timestamp/region.
  if (!gridMatched) {
    return {
      outcome: 'claim_withheld',
      limitation: 'grid_mismatch',
      varied,
      gridMatched,
      continuous,
      totalVariation: tv,
      summary:
        'trace drifts but does not match the grid reference within tolerance — no timestamp/region',
    };
  }

  // Continuity gate: a matched, varied trace with a splice is present but
  // tamper-suspect, so it is reported, never certified.
  if (!continuous) {
    return {
      outcome: 'observation_only',
      limitation: 'tamper_suspected',
      varied,
      gridMatched,
      continuous,
      totalVariation: tv,
      summary:
        'trace matches the grid but a successive jump exceeds the step bound — splice suspected, not certified',
    };
  }

  return {
    outcome: 'claim_admitted',
    limitation: 'none',
    varied,
    gridMatched,
    continuous,
    totalVariation: tv,
    summary: `varied trace (total variation ${tv}) matches the grid and is continuous — timestamp/grid certified`,
  };
}

/* ------------------------------------------------------------------ */
/*  Finite certificate                                                */
/* ------------------------------------------------------------------ */

/**
 * A finite certificate that Monster Mesh has all required ENF gates.
 * Mirrors `ElectricalNetworkFrequency.ENFCertificate`.
 */
export interface ENFCertificate {
  readonly traceVaried: boolean;
  readonly gridMatched: boolean;
  readonly uniquenessRefuted: boolean;
  readonly continuityChecked: boolean;
  readonly tamperRejected: boolean;
}

/** The ENF carrier for Monster Mesh promotion checks.
 *  Mirrors `ElectricalNetworkFrequency.enfRusticCertificate`. */
export const enfRusticCertificate: ENFCertificate = {
  traceVaried: true,
  gridMatched: true,
  uniquenessRefuted: true,
  continuityChecked: true,
  tamperRejected: true,
};

/* ------------------------------------------------------------------ */
/*  Concrete fixtures (mirror the Lean fixtures)                      */
/* ------------------------------------------------------------------ */

/** The grid's logged excursion sequence over the window (region G1). A genuine
 *  drift. Mirrors `ElectricalNetworkFrequency.gridRef`. */
export const gridRef: Trace = [50, 51, 49, 51, 48, 50];

/** A recording whose recovered ENF trace matches `gridRef` within tolerance 1.
 *  Timestamps + geo-locates to grid G1. Mirrors `matchedTrace`. */
export const matchedTrace: Trace = [50, 50, 50, 51, 49, 50];

/** A flat trace: the mains were absent / the recording was synthesized, so the
 *  trace is constant. No drift, no timestamp. Mirrors `constantTrace`. */
export const constantTrace: Trace = [50, 50, 50, 50, 50, 50];

/** A second grid region (G2) whose reference is well separated from G1 (at
 *  index 0 they differ by `60 - 50 = 10 > 2*tol`). The hard negative.
 *  Mirrors `otherGrid`. */
export const otherGrid: Trace = [60, 61, 59, 62, 58, 60];

/** A spliced recording: continuous up to the cut, then a large discontinuity
 *  (`51 → 70`, a jump of 19) where a second clip was pasted in.
 *  Mirrors `splicedTrace`. */
export const splicedTrace: Trace = [50, 51, 49, 52, 70, 50];

/**
 * Electrical Network Frequency master witness. Bundles the boundary facts the
 * Lean theorem `enf_master` proves, each recomputed here from first principles
 * (never imported from Lean IR).
 *
 * Mirrors `ElectricalNetworkFrequency.enf_master`.
 */
export const ENFMaster = {
  /** The matched trace drifts: it passes the variation gate. */
  matchedTraceVaried: variationAdmits(matchedTrace) === true,

  /** The constant trace is rejected by the variation gate. */
  constantTraceRejected: variationAdmits(constantTrace) === false,

  /** The varied recording matches grid G1 within tolerance 1. */
  matchedTraceMatchesGrid: matches(1, matchedTrace, gridRef) === true,

  /** The matched trace does NOT match the well-separated grid G2 (hard negative). */
  matchedTraceNotOtherGrid: matches(1, matchedTrace, otherGrid) === false,

  /** The two grids really are separated by more than `2*tol` at index 0
   *  (`adist 50 60 = 10 > 2`), making the refutation non-vacuous. */
  gridsWellSeparated: adist(50, 60) > 2 * 1,

  /** The genuine grid reference is continuous (not flagged) at step bound 3. */
  gridRefNotFlagged: tamperFlagged(3, gridRef) === false,

  /** The spliced trace is flagged at step bound 3 (the `51 → 70` jump). */
  splicedTraceFlagged: tamperFlagged(3, splicedTrace) === true,

  /** The certificate records trace variation. */
  certTraceVaried: enfRusticCertificate.traceVaried,

  /** The certificate records uniqueness refutation. */
  certUniquenessRefuted: enfRusticCertificate.uniquenessRefuted,

  /** The certificate records tamper rejection. */
  certTamperRejected: enfRusticCertificate.tamperRejected,
} as const;
