// @a0n/gnosis/distributed-inference-host/src/english-frequency.ts
// English language statistics, frozen reference tables. Used by the cipher
// sieve (text-cipher-sieve.ts) as the natural-English distance metric.
// Standard cryptanalysis tables; the monogram values are Cornell-derived;
// bigram and quadgram values are rough representative log10 frequencies
// (full quadgram dataset is ~389k entries — the embedded subset is just
// the load-bearing top-50). For a production cipher breaker, swap the
// embedded tables for the full Cornell quadgram dataset; the load shape
// stays the same.

/** Normalize a string to A-Z uppercase (drop non-letters). */
export function normalize(text: string): string {
  let out = "";
  for (let i = 0; i < text.length; i++) {
    const c = text.charCodeAt(i);
    // A-Z
    if (c >= 0x41 && c <= 0x5a) {
      out += text.charAt(i);
    } else if (c >= 0x61 && c <= 0x7a) {
      // a-z -> A-Z
      out += String.fromCharCode(c - 0x20);
    }
  }
  return out;
}

// --- Monograms ----------------------------------------------------------

// Cornell English letter frequencies (A..Z).
const MONOGRAM_FREQ: readonly number[] = Object.freeze([
  0.0817, 0.015, 0.0278, 0.0425, 0.127, 0.0223, 0.0202, 0.0609, 0.0697,
  0.0015, 0.0077, 0.0403, 0.0241, 0.0675, 0.0751, 0.0193, 0.001, 0.0599,
  0.0633, 0.0906, 0.0276, 0.0098, 0.0236, 0.0015, 0.0197, 0.0007,
]);

/** log10 monogram probability (length 26 array, A=0, B=1, ..., Z=25). */
export const ENGLISH_MONOGRAM_LOG10: readonly number[] = Object.freeze(
  MONOGRAM_FREQ.map((f) => Math.log10(f)),
);

// --- Bigrams ------------------------------------------------------------

const BIGRAM_ENTRIES: ReadonlyArray<readonly [string, number]> = Object.freeze([
  ["TH", -1.40], ["HE", -1.46], ["IN", -1.62], ["ER", -1.66], ["AN", -1.68],
  ["RE", -1.74], ["ON", -1.79], ["AT", -1.82], ["EN", -1.87], ["ND", -1.89],
  ["TI", -1.91], ["ES", -1.93], ["OR", -1.94], ["TE", -1.97], ["OF", -1.98],
  ["ED", -2.00], ["IS", -2.01], ["IT", -2.04], ["AL", -2.06], ["AR", -2.07],
  ["ST", -2.10], ["TO", -2.10], ["NT", -2.12], ["NG", -2.14], ["SE", -2.16],
  ["HA", -2.17], ["AS", -2.20], ["OU", -2.22], ["IO", -2.23], ["LE", -2.25],
  ["VE", -2.27], ["CO", -2.30], ["ME", -2.32], ["DE", -2.33], ["HI", -2.34],
  ["RI", -2.36], ["RO", -2.38], ["IC", -2.40], ["NE", -2.42], ["EA", -2.44],
  ["RA", -2.46], ["CE", -2.48], ["LI", -2.50], ["CH", -2.52], ["LL", -2.54],
  ["BE", -2.56], ["MA", -2.58], ["SI", -2.60], ["OM", -2.62], ["UR", -2.64],
]);

/** log10 bigram probability (Map from 'TH', 'HE', ...). */
export const ENGLISH_BIGRAM_LOG10: ReadonlyMap<string, number> = new Map(
  BIGRAM_ENTRIES,
);

/** Floor log-probability for unknown bigrams. */
export const ENGLISH_BIGRAM_FLOOR_LOG10: number = -5.0;

// --- Quadgrams ----------------------------------------------------------

const QUADGRAM_ENTRIES: ReadonlyArray<readonly [string, number]> = Object.freeze([
  ["THER", -3.20], ["TION", -3.30], ["ATIO", -3.40], ["NTHE", -3.45],
  ["OFTH", -3.50], ["INGS", -3.55], ["IONS", -3.60], ["EVER", -3.65],
  ["ATER", -3.68], ["ANDT", -3.70], ["NTER", -3.72], ["OTHE", -3.74],
  ["HERE", -3.76], ["FROM", -3.78], ["INTO", -3.80], ["ESTH", -3.82],
  ["TICS", -3.84], ["ITHE", -3.86], ["FORM", -3.88], ["WHER", -3.90],
  ["EARS", -3.92], ["MENT", -3.94], ["AINT", -3.96], ["ALLY", -3.98],
  ["COUR", -4.00], ["LITT", -4.02], ["PEOP", -4.04], ["OUGH", -4.06],
  ["ENCE", -4.08], ["ESTE", -4.10], ["RENT", -4.12], ["OUSE", -4.14],
  ["ONLY", -4.16], ["ABOU", -4.18], ["INTH", -4.20], ["ITHA", -4.22],
  ["ATHE", -4.24], ["ELIK", -4.28], ["EAST", -4.30], ["WITH", -4.32],
  ["BEEN", -4.34], ["BOUT", -4.36], ["TURN", -4.38], ["AFTE", -4.40],
  ["OVER", -4.42], ["ETHA", -4.44], ["NTOF", -4.46], ["ENTL", -4.48],
  ["RTHE", -4.50], ["WOUL", -4.52],
]);

/** log10 quadgram probability (Map from 'THER', 'IONS', ...). */
export const ENGLISH_QUADGRAM_LOG10: ReadonlyMap<string, number> = new Map(
  QUADGRAM_ENTRIES,
);

/** Floor log-probability for unknown quadgrams. */
export const ENGLISH_QUADGRAM_FLOOR_LOG10: number = -8.0;

// --- Scoring functions --------------------------------------------------

/** Score a candidate plaintext by quadgram log-likelihood.
 *  Higher = more English-like. Returns the AVERAGE quadgram log-prob. */
export function quadgramScore(text: string): number {
  const norm = normalize(text);
  if (norm.length === 0) {
    throw new RangeError("quadgramScore: empty input after normalization");
  }
  if (norm.length < 4) {
    // No quadgrams fit; every window is a miss => floor.
    return ENGLISH_QUADGRAM_FLOOR_LOG10;
  }
  let sum = 0;
  let count = 0;
  for (let i = 0; i <= norm.length - 4; i++) {
    const q = norm.substring(i, i + 4);
    const v = ENGLISH_QUADGRAM_LOG10.get(q);
    sum += v !== undefined ? v : ENGLISH_QUADGRAM_FLOOR_LOG10;
    count++;
  }
  return sum / count;
}

/** Score by bigram log-likelihood (cheaper, less accurate). */
export function bigramScore(text: string): number {
  const norm = normalize(text);
  if (norm.length === 0) {
    throw new RangeError("bigramScore: empty input after normalization");
  }
  if (norm.length < 2) {
    return ENGLISH_BIGRAM_FLOOR_LOG10;
  }
  let sum = 0;
  let count = 0;
  for (let i = 0; i <= norm.length - 2; i++) {
    const b = norm.substring(i, i + 2);
    const v = ENGLISH_BIGRAM_LOG10.get(b);
    sum += v !== undefined ? v : ENGLISH_BIGRAM_FLOOR_LOG10;
    count++;
  }
  return sum / count;
}

/** Score by monogram-only chi-squared from English (cheap, coarse). */
export function chiSquaredEnglishDistance(text: string): number {
  const norm = normalize(text);
  if (norm.length === 0) {
    throw new RangeError(
      "chiSquaredEnglishDistance: empty input after normalization",
    );
  }
  const counts = new Array<number>(26).fill(0);
  for (let i = 0; i < norm.length; i++) {
    const idx = norm.charCodeAt(i) - 0x41;
    counts[idx]++;
  }
  const N = norm.length;
  let chi2 = 0;
  for (let i = 0; i < 26; i++) {
    const expected = N * MONOGRAM_FREQ[i];
    if (expected <= 0) continue;
    const diff = counts[i] - expected;
    chi2 += (diff * diff) / expected;
  }
  return chi2;
}

/** Index of Coincidence: ~0.067 for English, ~0.0385 for random. Used to
 *  detect Caesar (preserves IC) vs Vigenère (lowers IC) and to guess
 *  Vigenère key length. */
export function indexOfCoincidence(text: string): number {
  const norm = normalize(text);
  if (norm.length === 0) {
    throw new RangeError(
      "indexOfCoincidence: empty input after normalization",
    );
  }
  const N = norm.length;
  if (N < 2) return 0;
  const counts = new Array<number>(26).fill(0);
  for (let i = 0; i < N; i++) {
    const idx = norm.charCodeAt(i) - 0x41;
    counts[idx]++;
  }
  let sum = 0;
  for (let i = 0; i < 26; i++) {
    const n = counts[i];
    sum += n * (n - 1);
  }
  return sum / (N * (N - 1));
}

/** Combined n-gram English-likeness score. Blends monogram chi-squared
 *  (length-normalized), bigram log-likelihood, and quadgram log-likelihood
 *  with empirical weights (0.6 quad + 0.3 bi + 0.1 mono).
 *
 *  Higher = more English. Useful when the search landscape needs a
 *  smoother gradient than pure quadgram — the Z408 pair-swap plateau
 *  behaviour is partly because pure-quadgram is constant across many
 *  permutations of equally-improbable mappings; the lower-order n-grams
 *  resolve those ties.
 *
 *  Normalization rationale: chi² grows with text length; we divide by
 *  length to keep the monogram term comparable to the per-token bigram
 *  and quadgram log-likelihoods. */
export function combinedEnglishScore(text: string): number {
  if (text.length === 0) {
    throw new RangeError(
      'combinedEnglishScore: text must contain at least one character',
    );
  }
  const monoChi2 = chiSquaredEnglishDistance(text);
  const monoNorm = -monoChi2 / Math.max(1, text.length);
  const bi = text.length >= 2 ? bigramScore(text) : -5.0;
  const quad = text.length >= 4 ? quadgramScore(text) : -8.0;
  return 0.6 * quad + 0.3 * bi + 0.1 * monoNorm;
}
