// @a0n/gnosis/distributed-inference-host/src/footstep-gait-leak.ts
/**
 * # Footstep Gait Leak Detector — a privacy WARNING carrier
 *
 * Mirrors `Gnosis.Mesh.FootstepGaitLeak` from `open-source/gnosis-math`, the
 * Rustic Church carrier for the footstep-acoustics privacy hazard framed
 * **strictly defensively** as a leak detector that *warns a user*, never as a
 * person-identification tool. The physical fact it models: a person's footstep
 * cadence and rhythm signature — stride period, double-support time, impact
 * sharpness, gait asymmetry — is acoustically distinctive, so a nearby
 * microphone can LEAK *who is walking* (re-identification). This module does NOT
 * identify anyone and carries no machinery to do so; it only FLAGS that the
 * environment is acoustically revealing of gait, so a user can be warned that
 * their walking is exposed to re-identification.
 *
 * The polarity is the INVERSE of `./microphone-fingerprint`: a same-device
 * provenance match wants two signatures to be CLOSE; the gait leak fires when
 * two DIFFERENT people's gaits are FAR apart, because that separability is
 * exactly what an eavesdropper could exploit to tell walkers apart. Here
 * "admitted" means a privacy LEAK is admitted as present — gait is
 * re-identifiable, warn the user — NOT that any person was identified. Directly
 * mirrors `./keystroke-leak-detector`, with keys replaced by people's gaits.
 *
 * ## What is (and is not) carried
 *
 * This module is the runtime promotion check for a gait-leak privacy warning. It
 * carries ONLY the finite contract the Lean file proves:
 *
 *  - the saturating per-feature gait metric (`adist`);
 *  - the signature distance as the summed pointwise `adist`, with a large
 *    sentinel on a length mismatch (`fpDistance` / `mismatchPenalty`);
 *  - the leak gate as separation beyond `sepThreshold` (`gaitLeaks`);
 *  - the safe gate as indistinguishability within `intraTol`;
 *  - the leak-warning verdict folded into the epistemic-outcome vocabulary
 *    (`assessGaitLeak`).
 *
 * Real cadence extraction, footstep onset detection, floor/shoe coloration, room
 * reverberation, microphone placement, and the actual statistical separability
 * of a population's gaits stay runtime/empirical obligations OUTSIDE this
 * certificate — they are not theorems and are not asserted here.
 *
 * The promotion check folds the contract into the same epistemic-outcome
 * vocabulary the sibling mesh carriers already speak
 * (`claim_admitted` / `observation_only` / `claim_withheld`), so a "your walk
 * may be acoustically re-identifiable here" warning is raised only when the
 * Lean-side separability gate fires. STRICTLY DEFENSIVE throughout: this carrier
 * neither performs nor enables person identification.
 *
 * ## Lean source-of-truth
 *
 * @see open-source/gnosis-math/Gnosis/Mesh/FootstepGaitLeak.lean
 * @see open-source/gnosis-math/Gnosis/Mesh/KeystrokeLeakDetector.lean (sibling leak carrier)
 * @see open-source/gnosis-math/Gnosis/Mesh/MicrophoneFingerprint.lean (inverse-polarity sibling)
 */

/* ------------------------------------------------------------------ */
/*  Per-feature distance and gait signatures                          */
/* ------------------------------------------------------------------ */

/**
 * A gait signature is the per-feature acoustic signature of one walk: a list of
 * quantized rhythm features, e.g.
 * `[stridePeriod, doubleSupportTime, impactSharpness, asymmetry]`. Features are
 * non-negative integers (saturating `Nat` arithmetic), matching the Lean
 * `GaitSignature`.
 */
export type GaitSignature = readonly number[];

/** Absolute difference on `Nat` via saturating subtraction (`|a - b|`). The
 *  per-feature gait residual metric, shared with the mesh corridor carriers.
 *  Mirrors `FootstepGaitLeak.adist`. */
export function adist(a: number, b: number): number {
  return Math.max(0, a - b) + Math.max(0, b - a);
}

/**
 * Sentinel returned on a length mismatch: large enough that no sane threshold
 * treats it as a near match. Differently-windowed gait signatures are reported
 * as a large separation. Mirrors `FootstepGaitLeak.mismatchPenalty`.
 */
export const mismatchPenalty = 1000000;

/**
 * The signature distance: the sum of pointwise `adist` between aligned gait
 * features. On a length mismatch it returns `mismatchPenalty` rather than
 * attempting a partial alignment. Mirrors `FootstepGaitLeak.fpDistance`.
 */
export function fpDistance(x: GaitSignature, y: GaitSignature): number {
  if (x.length !== y.length) {
    return mismatchPenalty;
  }
  let sum = 0;
  for (let i = 0; i < x.length; i++) {
    sum += adist(x[i], y[i]);
  }
  return sum;
}

/* ------------------------------------------------------------------ */
/*  Leak gate (inter-person separability — THE WARNING)               */
/* ------------------------------------------------------------------ */

/**
 * The privacy-leak warning predicate. Two people's gait signatures are
 * *separable* — hence the environment LEAKS who is walking (re-identification) —
 * when their signature distance exceeds `sepThreshold`. This is the inverse of a
 * fingerprint match: separation, not closeness, is the hazard, because
 * separability is exactly what lets a nearby microphone tell two walkers apart.
 * Returns `true` to FLAG a leak (warn the user); it does NOT identify the
 * walker. Mirrors `FootstepGaitLeak.gaitLeaksB` / `GaitLeakPresent`.
 *
 * @param gaitX one walk's gait signature
 * @param gaitY another walk's gait signature
 * @param opts the separability threshold the distance must exceed to be a leak
 */
export function gaitLeaks(
  gaitX: GaitSignature,
  gaitY: GaitSignature,
  opts: { readonly sepThreshold: number },
): boolean {
  return fpDistance(gaitX, gaitY) > opts.sepThreshold;
}

/* ------------------------------------------------------------------ */
/*  Runtime promotion check (epistemic outcome)                       */
/* ------------------------------------------------------------------ */

/**
 * Epistemic outcome of a gait-leak privacy assessment. Identical vocabulary to
 * the sibling mesh carriers (`./keystroke-leak-detector`,
 * `./microphone-fingerprint`), kept as a local literal union so this certificate
 * has no import coupling. Note the polarity: `claim_admitted` here means a
 * privacy LEAK is admitted as present (warn the user), NOT that any person was
 * identified.
 *
 *  - `claim_admitted`    — the two gait signatures separate beyond `sepThreshold`,
 *                          so the walkers are acoustically distinguishable: a
 *                          re-identification LEAK is present and the user should
 *                          be warned.
 *  - `observation_only`  — the separation sits in the ambiguous band (above
 *                          `intraTol`, but not beyond `sepThreshold`): possibly
 *                          re-identifiable, not certifiably so — reported, never
 *                          promoted to a warning.
 *  - `claim_withheld`    — the signatures are within `intraTol`, i.e.
 *                          indistinguishable: no re-identification, no leak.
 */
export type GaitLeakEpistemicOutcome =
  | 'claim_admitted'
  | 'observation_only'
  | 'claim_withheld';

/** The reason a leak warning was held back (or the safe verdict reached). */
export type GaitLeakLimitationClass =
  | 'indistinguishable_safe'
  | 'ambiguous_margin'
  | 'none';

export interface GaitLeakAssessment {
  readonly outcome: GaitLeakEpistemicOutcome;
  readonly limitation: GaitLeakLimitationClass;
  /** Whether the separability gate fired (a re-identification leak is present). */
  readonly leaks: boolean;
  /** Whether the pair is within `intraTol` (acoustically indistinguishable). */
  readonly safe: boolean;
  /** The measured signature distance (the displayable separation). */
  readonly distance: number;
  readonly summary: string;
}

export interface GaitLeakAssessOptions {
  /** Intra-person tightness: at or below this distance the pair is indistinguishable. */
  readonly intraTol: number;
  /** Separability: beyond this distance the gaits leak (are re-identifiable). */
  readonly sepThreshold: number;
}

/**
 * The promotion check Monster Mesh runs before raising a "your walk may be
 * acoustically re-identifiable here" privacy notice. It folds the Lean contract
 * into an epistemic outcome, with leak-detector polarity:
 *
 *  - separation beyond `sepThreshold` is the LEAK signal — the gaits are
 *    distinguishable, so the warning is admitted (mirrors `distinct_gaits_leak`);
 *  - a pair within `intraTol` is indistinguishable — no re-identification,
 *    withheld as SAFE (mirrors `indistinct_gaits_safe` / `same_person_not_leak`);
 *  - the band in between is an ambiguous margin: reported as an observation, not
 *    promoted to a warning.
 *
 * This neither identifies nor re-identifies any person; it only flags whether
 * the environment is acoustically revealing of gait.
 *
 * @param gaitX one walk's gait signature
 * @param gaitY another walk's gait signature
 * @param opts the intra-person tolerance and the separability threshold
 */
export function assessGaitLeak(
  gaitX: GaitSignature,
  gaitY: GaitSignature,
  opts: GaitLeakAssessOptions,
): GaitLeakAssessment {
  const distance = fpDistance(gaitX, gaitY);
  const leaks = distance > opts.sepThreshold;
  const safe = distance <= opts.intraTol;

  // Leak gate: separability beyond the threshold means the walkers are
  // distinguishable — a re-identification leak is present, warn the user.
  if (leaks) {
    return {
      outcome: 'claim_admitted',
      limitation: 'none',
      leaks,
      safe,
      distance,
      summary: `gaits acoustically distinguishable (distance ${distance} > sepThreshold ${opts.sepThreshold}) — re-identification LEAK present, warn the user`,
    };
  }

  // Safe gate: within the intra-person spread the gaits are indistinguishable,
  // so an eavesdropper cannot tell the walkers apart — no re-identification.
  if (safe) {
    return {
      outcome: 'claim_withheld',
      limitation: 'indistinguishable_safe',
      leaks,
      safe,
      distance,
      summary: `gaits indistinguishable (distance ${distance} <= intraTol ${opts.intraTol}) — no leak, safe`,
    };
  }

  // Ambiguous margin: above the intra-person spread but not beyond the leak
  // threshold — possibly re-identifiable, not certifiably so.
  return {
    outcome: 'observation_only',
    limitation: 'ambiguous_margin',
    leaks,
    safe,
    distance,
    summary: `separation ${distance} sits between intraTol ${opts.intraTol} and sepThreshold ${opts.sepThreshold} — ambiguous, reported but not promoted to a warning`,
  };
}

/* ------------------------------------------------------------------ */
/*  Concrete fixtures (mirror the Lean fixtures)                      */
/* ------------------------------------------------------------------ */

/** Walk 1 of person A: its measured per-feature gait signature (length 4)
 *  `[stridePeriod, doubleSupportTime, impactSharpness, asymmetry]`.
 *  Mirrors `FootstepGaitLeak.walkA1`. */
export const walkA1: GaitSignature = [50, 18, 12, 4];

/** Walk 2 of the SAME person A: near-identical, off by at most one unit per
 *  feature. `fpDistance(walkA1, walkA2) = 1+1+0+1 = 3`, tight within
 *  `intraTol = 4` and far below `sepThreshold = 12`, so the two walks are NOT
 *  flagged as a leak between them. Mirrors `walkA2`. */
export const walkA2: GaitSignature = [51, 17, 12, 5];

/** A clearly DIFFERENT person B: their gait signature is far from person A's.
 *  `fpDistance(walkA1, walkB) = 12+12+23+18 = 65`, well beyond `sepThreshold = 12`,
 *  so the pair is separable and the environment LEAKS who is walking
 *  (re-identification flagged). Mirrors `walkB`. */
export const walkB: GaitSignature = [38, 30, 35, 22];

/** A degenerate, acoustically near-identical coarse gait, part 1.
 *  Mirrors `coarseP`. */
export const coarseP: GaitSignature = [25, 20, 10, 6];

/** Its near-identical neighbour: `fpDistance(coarseP, coarseQ) = 0+0+0+1 = 1`,
 *  within `intraTol = 4` and far below `sepThreshold = 12`. Indistinguishable →
 *  SAFE → not flagged. Mirrors `coarseQ`. */
export const coarseQ: GaitSignature = [25, 20, 10, 7];

/**
 * Footstep Gait Leak master witness. Bundles the boundary facts the Lean theorem
 * `gait_leak_master` proves, each recomputed here from first principles (never
 * imported from Lean IR).
 *
 * Mirrors `FootstepGaitLeak.gait_leak_master`.
 */
export const FootstepGaitLeakMaster = {
  /** A signature is at zero distance from itself (self-consistency limit). */
  fpDistanceSelfZero: fpDistance(walkA1, walkA1) === 0,

  /** The two walks of person A cluster tightly within tolerance 4 (distance 3). */
  samePersonTight: fpDistance(walkA1, walkA2) <= 4,

  /** Person A and person B separate by more than sepThreshold 12 (distance 65). */
  gaitsWellSeparated: fpDistance(walkA1, walkB) > 12,

  /** The separated pair triggers the re-identification leak warning. */
  distinctGaitsLeak:
    assessGaitLeak(walkA1, walkB, { intraTol: 4, sepThreshold: 12 })
      .outcome === 'claim_admitted',

  /** The Boolean detector also fires on the separated pair. */
  distinctGaitsLeakB: gaitLeaks(walkA1, walkB, { sepThreshold: 12 }) === true,

  /** The same-person pair is NOT a leak (no separation between two walks). */
  samePersonNotLeak: gaitLeaks(walkA1, walkA2, { sepThreshold: 12 }) === false,

  /** The near-identical coarse gaits are indistinguishable within tolerance 4 (distance 1). */
  indistinctGaitsSafe:
    assessGaitLeak(coarseP, coarseQ, { intraTol: 4, sepThreshold: 12 })
      .outcome === 'claim_withheld',

  /** And the detector does NOT fire on the indistinguishable pair. */
  indistinctGaitsNotFlagged:
    gaitLeaks(coarseP, coarseQ, { sepThreshold: 12 }) === false,
} as const;
