// @a0n/gnosis/distributed-inference-host/src/frf.ts
// Generic FORK/RACE/FOLD primitive. FORK launches an array of async
// tasks concurrently (with optional concurrency cap so we don't
// oversubscribe a single-core JS host on CPU-bound work). RACE is a
// filter step — surviving results pass through, others are dropped.
// FOLD merges survivors into a single output. Mirrors the pattern in
// @a0n/aether/src/glossolalia-moa.ts stripped of MOA-specific
// token/agent machinery, and the spirit (not implementation) of
// @a0n/gnosis-frf's static rayon replacement. Throws if every fork
// rejects or every result is filtered out — the caller can distinguish
// "all forks failed" from "no survivors" via the error code.

import { Verdict } from '@a0n/bitwise/triton-verdict';

export type ForkTask<T> = () => Promise<T>;

export interface ForkRaceFoldArgs<T, U> {
  /** N parallel tasks; FORK launches them. */
  forks: ReadonlyArray<ForkTask<T>>;
  /** RACE step: keep only survivors. Default: identity (keep all completed). */
  race?: (results: ReadonlyArray<T>) => ReadonlyArray<T>;
  /** FOLD step: merge survivors into a single output. */
  fold: (survivors: ReadonlyArray<T>) => U;
  /** Max in-flight forks. Default: forks.length (no cap). */
  concurrency?: number;
  /** When true, a single fork rejection aborts the run. Default false (rejected
   *  forks are dropped before RACE). */
  abortOnReject?: boolean;
}

export interface ForkRaceFoldReport {
  forkCount: number;
  fulfilledCount: number;
  rejectedCount: number;
  survivorCount: number;
}

export interface ForkRaceFoldResult<U> {
  output: U;
  report: ForkRaceFoldReport;
  /**
   * The 3-state FORK verdict (canonical TritonVerdict, replacing the old
   * private `sawRejection: boolean`). The boolean could only say "clean" or
   * "something rejected" and silently squashed the partial-but-recoverable
   * case into the rejection bucket. The verdict restores the neutral middle:
   *
   *   - `Accept`  — every fork fulfilled (`rejectedCount === 0`): full success.
   *   - `Abstain` — some forks rejected but survivors remain
   *                 (`0 < rejectedCount < forkCount`): partial / budget-
   *                 exhausted-but-recoverable. DEFER — the fold still produced
   *                 an output, but a later fold/retry could fill the gaps. This
   *                 is the value-add the boolean could not express: it forced
   *                 this case to read as a rejection.
   *   - `Decline` — every fork rejected (`rejectedCount === forkCount`): a hard
   *                 rejection. (This is the same condition under which the run
   *                 throws `all forks rejected`; on the non-throwing path the
   *                 verdict names the cause at the all-rejected extreme.)
   *
   * Bool-subsumption (preserved behavior): the legacy `sawRejection === true`
   * is exactly "any fork rejected" and still drives `abortOnReject`. At the
   * all-rejected extreme `isReject(verdict)` ≅ old `sawRejection === true`; for
   * the partial case the bool could only see `true`, where the verdict now
   * defers with `Abstain` — the information the bool could not hold. Lives on
   * the result (not the `report`) so the report's existing exact shape is
   * preserved for callers that deep-equal it.
   */
  verdict: Verdict;
}

/** Generic FORK/RACE/FOLD over async tasks.
 *
 * @throws RangeError if forks is empty.
 * @throws Error('forkRaceFold: all forks rejected') if every fork rejects.
 * @throws Error('forkRaceFold: no survivors after race') if RACE filters
 *         out everything that succeeded.
 */
export async function forkRaceFold<T, U>(
  args: ForkRaceFoldArgs<T, U>,
): Promise<ForkRaceFoldResult<U>> {
  const { forks, race, fold, concurrency, abortOnReject } = args;

  if (forks.length === 0) {
    throw new RangeError('forkRaceFold: forks must be non-empty');
  }

  const n = forks.length;
  const cap =
    concurrency === undefined || concurrency >= n
      ? n
      : Math.max(1, concurrency | 0);

  // Pre-allocate slot-aligned settled array so order matches fork-input order.
  const settled: Array<PromiseSettledResult<T>> = new Array(n);

  if (cap >= n) {
    // No cap: fan everything out at once.
    const all = await Promise.allSettled(forks.map((f) => f()));
    for (let i = 0; i < n; i++) settled[i] = all[i];
  } else {
    // Capped: shared cursor, `cap` worker loops pull indices.
    let cursor = 0;
    const worker = async (): Promise<void> => {
      while (true) {
        const i = cursor++;
        if (i >= n) return;
        try {
          const value = await forks[i]();
          settled[i] = { status: 'fulfilled', value };
        } catch (reason) {
          settled[i] = { status: 'rejected', reason };
        }
      }
    };
    const workers: Array<Promise<void>> = [];
    for (let w = 0; w < cap; w++) workers.push(worker());
    await Promise.all(workers);
  }

  // Partition while preserving fork-input order.
  const fulfilled: T[] = [];
  let rejectedCount = 0;
  let firstRejection: unknown = undefined;
  let sawRejection = false;
  for (let i = 0; i < n; i++) {
    const r = settled[i];
    if (r.status === 'fulfilled') {
      fulfilled.push(r.value);
    } else {
      rejectedCount++;
      if (!sawRejection) {
        firstRejection = r.reason;
        sawRejection = true;
      }
    }
  }

  // Collapse the fork run to a canonical 3-state verdict. `sawRejection`
  // (any fork rejected) is preserved verbatim as the `abortOnReject` driver,
  // and `isReject(forkVerdict)` ≅ `sawRejection === true` only at the
  // all-rejected extreme — the partial case defers (`Abstain`) instead, which
  // the old boolean could not express.
  const forkVerdict: Verdict =
    rejectedCount === 0
      ? Verdict.Accept // every fork fulfilled — full success.
      : rejectedCount === n
        ? Verdict.Decline // every fork rejected — hard rejection.
        : Verdict.Abstain; // partial / recoverable — defer for a later fold.

  // Preserved behavior: the legacy `sawRejection` boolean (any rejection)
  // still drives `abortOnReject`. `isReject(forkVerdict)` holds only at the
  // all-rejected extreme (Decline) and so is a strict subset of `sawRejection`;
  // we keep `sawRejection` as the literal abort trigger to honor the existing
  // partial-abort semantics exactly. (`isReject(forkVerdict) ⇒ sawRejection`.)
  if (abortOnReject && sawRejection) {
    throw firstRejection instanceof Error
      ? firstRejection
      : new Error(String(firstRejection));
  }

  const raced =
    race === undefined ? fulfilled : Array.from(race(fulfilled));

  if (raced.length === 0) {
    if (fulfilled.length === 0) {
      throw new Error('forkRaceFold: all forks rejected');
    }
    throw new Error('forkRaceFold: no survivors after race');
  }

  const output = fold(raced);
  return {
    output,
    report: {
      forkCount: n,
      fulfilledCount: fulfilled.length,
      rejectedCount,
      survivorCount: raced.length,
    },
    verdict: forkVerdict,
  };
}

/** Convenience shorthand: identity RACE + caller-supplied FOLD. */
export async function forkFold<T, U>(
  forks: ReadonlyArray<ForkTask<T>>,
  fold: (results: ReadonlyArray<T>) => U,
  concurrency?: number,
): Promise<U> {
  const { output } = await forkRaceFold<T, U>({
    forks,
    fold,
    concurrency,
  });
  return output;
}
