// @a0n/gnosis/distributed-inference-host/src/griffin-lim.ts
// Griffin-Lim phase reconstruction (1984).
// Recovers a real audio signal from a magnitude-only STFT by alternating
// between iSTFT and re-STFT, keeping the target magnitude and adopting
// the projected phase each round. Used as the last stage of audio whisper
// recovery — after our 2D Wiener cleans the spectrogram magnitude, this
// rebuilds a coherent audio signal from the cleaned magnitude.

import { stft, istft, type STFTConfig } from './audio-stft';

export interface GriffinLimArgs {
  /** Target magnitude (length = nFrames * nFreqs, row-major). */
  magnitude: Float32Array;
  nFrames: number;
  config: STFTConfig;
  /** Iteration count. Default 30. Must be >= 1. */
  iterations?: number;
  /** Initial phase: 'zero' (deterministic), 'random' (LCG-seeded), or a Float32Array of explicit phases. Default 'zero'. */
  initialPhase?: 'zero' | 'random' | Float32Array;
  /** Seed for the random initial phase. Default 0xC0FFEE. */
  seed?: number;
}

export interface GriffinLimResult {
  audio: Float32Array;
  phase: Float32Array;
}

/**
 * Handles the gnosis griffin Lim workflow.
 */
export function griffinLim(args: GriffinLimArgs): GriffinLimResult {
  const { magnitude, nFrames, config } = args;
  const iterations = args.iterations ?? 30;
  const initialPhase = args.initialPhase ?? 'zero';

  if (!Number.isFinite(iterations) || iterations < 1) {
    throw new RangeError(
      `griffinLim: iterations must be >= 1, got ${iterations}`,
    );
  }

  const nFreqs = config.nFft / 2 + 1;
  const expectedLen = nFrames * nFreqs;
  if (magnitude.length !== expectedLen) {
    throw new RangeError(
      `griffinLim: magnitude.length (${magnitude.length}) must equal nFrames * (nFft/2+1) = ${nFrames} * ${nFreqs} = ${expectedLen}`,
    );
  }

  let phase = new Float32Array(magnitude.length);
  if (initialPhase === 'random') {
    let s = (args.seed ?? 0xc0ffee) >>> 0;
    const rand = () => {
      s = (s * 1103515245 + 12345) & 0x7fffffff;
      return s / 0x7fffffff;
    };
    for (let i = 0; i < magnitude.length; i++) {
      phase[i] = (rand() - 0.5) * 2 * Math.PI;
    }
  } else if (initialPhase instanceof Float32Array) {
    if (initialPhase.length !== magnitude.length) {
      // If the provided phase is shorter, copy what we can and leave the rest zero
      const len = Math.min(initialPhase.length, phase.length);
      phase.set(initialPhase.subarray(0, len));
    } else {
      phase.set(initialPhase);
    }
  }

  for (let it = 0; it < iterations; it++) {
    const signal = istft({ magnitude, phase, nFrames, config });
    const probe = stft(signal, config);
    // Adopt probe.phase but keep target magnitude.
    // istft/stft already return phase as atan2(im, re), so no recomputation needed.
    phase = probe.phase as any;
  }

  const audio = istft({ magnitude, phase, nFrames, config });
  return { audio, phase };
}
