// @a0n/gnosis/distributed-inference-host/src/hf-tokenizer-json.ts
import { readFileSync } from 'node:fs';

type HfAddedToken = {
  id: number;
  content: string;
  special?: boolean;
};

type HfTokenizerJson = {
  model?: {
    type?: string;
    vocab?: Record<string, number>;
    // tokenizers >=0.14 emits merges as ["a","b"] pairs; older versions emit
    // "a b" strings. Both are accepted (see normalizeMerge).
    merges?: Array<string | [string, string]>;
    byte_fallback?: boolean;
    unk_token?: string;
  };
  // SentencePiece/Gemma express the space→▁ metaspace substitution as a
  // Replace normalizer (+ a no-op Split pre_tokenizer) rather than a Metaspace
  // pre_tokenizer. Accept either form.
  normalizer?: {
    type?: string;
    pattern?: { String?: string };
    content?: string;
    normalizers?: Array<{ type?: string; pattern?: { String?: string }; content?: string }>;
  };
  pre_tokenizer?: {
    type?: string;
    replacement?: string;
    prepend_scheme?: string;
    split?: boolean;
  };
  added_tokens?: HfAddedToken[];
};

export type HfBpeMetaspaceTokenizer = {
  readonly name: string;
  readonly stopTokenIds: number[];
  encode(text: string, options?: { addBos?: boolean }): number[];
  decode(tokenIds: readonly number[], options?: { skipSpecial?: boolean }): string;
  decodeToken(tokenId: number): string;
};

const TEXT_ENCODER = new TextEncoder();
const TEXT_DECODER = new TextDecoder('utf-8', { fatal: false });

function assertRecord(value: unknown, label: string): asserts value is Record<string, unknown> {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    throw new Error(`${label} must be a JSON object`);
  }
}

function parseTokenizerJson(text: string): HfTokenizerJson {
  const parsed = JSON.parse(text) as unknown;
  assertRecord(parsed, 'tokenizer JSON');
  return parsed as HfTokenizerJson;
}

function pairKey(left: string, right: string): string {
  return `${left}\u0000${right}`;
}

function parseMerge(merge: string | [string, string]): [string, string] | null {
  // Newer HF tokenizers (Gemma, Llama-3, …) store merges as ["a","b"] arrays.
  if (Array.isArray(merge)) {
    return merge.length === 2 ? [merge[0], merge[1]] : null;
  }
  const split = merge.indexOf(' ');
  if (split < 0) return null;
  return [merge.slice(0, split), merge.slice(split + 1)];
}

/// Detect the SentencePiece/Gemma metaspace replacement char from a Replace
/// normalizer (possibly nested in a Sequence): a `{type:Replace, pattern:{String:" "},
/// content:"▁"}` rule maps spaces to ▁ exactly like a Metaspace pre_tokenizer.
function metaspaceFromNormalizer(
  normalizer: HfTokenizerJson['normalizer'],
): string | null {
  if (!normalizer) return null;
  const candidates =
    normalizer.type === 'Sequence' && Array.isArray(normalizer.normalizers)
      ? normalizer.normalizers
      : [normalizer];
  for (const rule of candidates) {
    if (
      rule &&
      rule.type === 'Replace' &&
      rule.pattern?.String === ' ' &&
      typeof rule.content === 'string'
    ) {
      return rule.content;
    }
  }
  return null;
}

function byteFallbackToken(byte: number): string {
  return `<0x${byte.toString(16).toUpperCase().padStart(2, '0')}>`;
}

function fallbackBytesToText(token: string): Uint8Array | null {
  const match = token.match(/^<0x([0-9A-Fa-f]{2})>$/);
  if (!match) return null;
  return new Uint8Array([parseInt(match[1]!, 16)]);
}

/**
 * Creates the Hf Bpe Metaspace Tokenizer.
 */
export function createHfBpeMetaspaceTokenizer(
  text: string,
  name = 'hf-bpe-metaspace',
): HfBpeMetaspaceTokenizer {
  const tokenizer = parseTokenizerJson(text);
  if (tokenizer.model?.type !== 'BPE') {
    throw new Error(`unsupported HF tokenizer model: ${tokenizer.model?.type ?? 'unknown'}`);
  }
  // Accept two equivalent metaspace encodings:
  //   1. Metaspace pre_tokenizer (Mistral/Llama SentencePiece export).
  //   2. Replace-normalizer space→▁ + Split/absent pre_tokenizer (Gemma 2/3/4,
  //      Llama-3 SentencePiece export). Functionally identical for BPE.
  const normalizerReplacement = metaspaceFromNormalizer(tokenizer.normalizer);
  const isMetaspacePreTokenizer = tokenizer.pre_tokenizer?.type === 'Metaspace';
  if (!isMetaspacePreTokenizer && normalizerReplacement === null) {
    throw new Error(
      `unsupported HF pre-tokenizer: ${
        tokenizer.pre_tokenizer?.type ?? 'unknown'
      } (and no space→metaspace Replace normalizer found)`,
    );
  }

  const vocab = tokenizer.model.vocab;
  if (!vocab || Object.keys(vocab).length === 0) {
    throw new Error('HF tokenizer JSON does not contain model.vocab');
  }
  const vocabulary: Record<string, number> = vocab;

  const idToToken: string[] = [];
  for (const [token, id] of Object.entries(vocabulary)) {
    idToToken[id] = token;
  }
  const unkToken = tokenizer.model.unk_token ?? '<unk>';
  const unkTokenId = vocabulary[unkToken] ?? 0;
  // BOS/EOS resolution. Mistral/Llama use <s>/</s>; Gemma/SentencePiece use
  // <bos>/<eos>. CRITICAL: Gemma's vocab ALSO contains <s>/</s> as ordinary
  // (non-special) tokens, so a plain vocab lookup picks the wrong ids. Prefer
  // ids that are flagged `special` in added_tokens, falling back to a vocab
  // lookup only when the tokenizer carries no special-token table.
  const specialContentToId = new Map<string, number>();
  for (const t of tokenizer.added_tokens ?? []) {
    if (t.special && t.content) specialContentToId.set(t.content, t.id);
  }
  const resolveSpecial = (...names: string[]): number | undefined => {
    for (const n of names) {
      const sp = specialContentToId.get(n);
      if (sp !== undefined) return sp;
    }
    for (const n of names) {
      const v = vocabulary[n];
      if (v !== undefined) return v;
    }
    return undefined;
  };
  const bosTokenId = resolveSpecial('<s>', '<bos>');
  const eosTokenId = resolveSpecial('</s>', '<eos>', '<end_of_turn>');
  const replacement =
    tokenizer.pre_tokenizer?.replacement ?? normalizerReplacement ?? '▁';
  // Metaspace pre_tokenizer carries an explicit prepend_scheme; the Gemma
  // normalizer form relies on add_dummy_prefix (a leading space before the
  // first piece), which SentencePiece always applies. Default to prepend on
  // the normalizer path so " Paris" → "▁Paris" matches the trained vocab.
  const prependFirst = isMetaspacePreTokenizer
    ? tokenizer.pre_tokenizer?.prepend_scheme === 'first'
    : true;
  const byteFallback = tokenizer.model.byte_fallback !== false;

  const mergeRanks = new Map<string, number>();
  for (let rank = 0; rank < (tokenizer.model.merges ?? []).length; rank++) {
    const parsedMerge = parseMerge(tokenizer.model.merges![rank]!);
    if (parsedMerge) mergeRanks.set(pairKey(parsedMerge[0], parsedMerge[1]), rank);
  }

  const specialTokens = (tokenizer.added_tokens ?? [])
    .filter((token) => token.special && token.content)
    .map((token) => ({ token: token.content, id: token.id }))
    .sort((a, b) => b.token.length - a.token.length);
  const specialTokenIds = new Set(specialTokens.map((token) => token.id));

  function metaspace(textChunk: string, atInputStart: boolean): string {
    const needsPrefix =
      prependFirst && atInputStart && !textChunk.startsWith(' ') && textChunk.length > 0;
    const normalized = needsPrefix ? ` ${textChunk}` : textChunk;
    return normalized.replaceAll(' ', replacement);
  }

  function initialParts(piece: string): string[] {
    const parts: string[] = [];
    for (const char of Array.from(piece)) {
      if (vocabulary[char] !== undefined) {
        parts.push(char);
        continue;
      }
      if (!byteFallback) {
        parts.push(unkToken);
        continue;
      }
      for (const byte of TEXT_ENCODER.encode(char)) {
        const fallbackToken = byteFallbackToken(byte);
        parts.push(vocabulary[fallbackToken] !== undefined ? fallbackToken : unkToken);
      }
    }
    return parts;
  }

  function bpe(piece: string): string[] {
    let parts = initialParts(piece);
    if (parts.length <= 1) return parts;

    while (parts.length > 1) {
      let bestRank = Number.POSITIVE_INFINITY;
      let bestLeft = '';
      let bestRight = '';

      for (let i = 0; i < parts.length - 1; i++) {
        const rank = mergeRanks.get(pairKey(parts[i]!, parts[i + 1]!));
        if (rank !== undefined && rank < bestRank) {
          bestRank = rank;
          bestLeft = parts[i]!;
          bestRight = parts[i + 1]!;
        }
      }

      if (!Number.isFinite(bestRank)) break;

      const merged: string[] = [];
      for (let i = 0; i < parts.length; i++) {
        if (i < parts.length - 1 && parts[i] === bestLeft && parts[i + 1] === bestRight) {
          merged.push(bestLeft + bestRight);
          i += 1;
        } else {
          merged.push(parts[i]!);
        }
      }
      parts = merged;
    }

    return parts;
  }

  function encodeOrdinary(textChunk: string, atInputStart: boolean): number[] {
    const encoded = metaspace(textChunk, atInputStart);
    if (!encoded) return [];
    const tokenIds: number[] = [];
    for (const token of bpe(encoded)) {
      tokenIds.push(vocabulary[token] ?? unkTokenId);
    }
    return tokenIds;
  }

  function encodeBody(textValue: string): number[] {
    const tokenIds: number[] = [];
    let cursor = 0;

    while (cursor < textValue.length) {
      const matchedSpecial = specialTokens.find(({ token }) =>
        textValue.startsWith(token, cursor)
      );
      if (matchedSpecial) {
        tokenIds.push(matchedSpecial.id);
        cursor += matchedSpecial.token.length;
        continue;
      }

      let nextSpecialIndex = textValue.length;
      for (const { token } of specialTokens) {
        const index = textValue.indexOf(token, cursor);
        if (index >= 0 && index < nextSpecialIndex) nextSpecialIndex = index;
      }

      tokenIds.push(
        ...encodeOrdinary(textValue.slice(cursor, nextSpecialIndex), cursor === 0),
      );
      cursor = nextSpecialIndex;
    }

    return tokenIds;
  }

  function decodeTokenStrings(tokenIds: readonly number[], skipSpecial: boolean): string {
    let textValue = '';
    let pendingBytes: number[] = [];

    const flushBytes = () => {
      if (pendingBytes.length === 0) return;
      textValue += TEXT_DECODER.decode(new Uint8Array(pendingBytes));
      pendingBytes = [];
    };

    for (const tokenId of tokenIds) {
      if (skipSpecial && specialTokenIds.has(tokenId)) continue;
      const token = idToToken[tokenId];
      if (!token) continue;
      const fallback = fallbackBytesToText(token);
      if (fallback) {
        pendingBytes.push(...fallback);
        continue;
      }
      flushBytes();
      textValue += token;
    }

    flushBytes();
    return textValue;
  }

  // Stop tokens: EOS plus any turn-terminator special (Gemma <end_of_turn>,
  // ChatML <|im_end|>). BOS is NOT a stop — including it truncates output at
  // the first token for templates that re-emit a leading marker.
  const turnEndId = specialContentToId.get('<end_of_turn>');
  const imEndId = specialContentToId.get('<|im_end|>');
  const stopTokenIds = [eosTokenId, turnEndId, imEndId].filter(
    (id): id is number => typeof id === 'number',
  );

  return {
    name,
    stopTokenIds,
    encode(textValue: string, options: { addBos?: boolean } = {}): number[] {
      const tokenIds = encodeBody(textValue);
      return options.addBos && typeof bosTokenId === 'number'
        ? [bosTokenId, ...tokenIds]
        : tokenIds;
    },
    decode(tokenIds: readonly number[], options: { skipSpecial?: boolean } = {}): string {
      const raw = decodeTokenStrings(tokenIds, options.skipSpecial ?? true);
      const spaces = raw.replaceAll(replacement, ' ');
      return spaces.startsWith(' ') ? spaces.slice(1) : spaces;
    },
    decodeToken(tokenId: number): string {
      return this.decode([tokenId], { skipSpecial: true });
    },
  };
}

/**
 * Loads the Hf Bpe Metaspace Tokenizer From File.
 */
export function loadHfBpeMetaspaceTokenizerFromFile(
  path: string,
  name = 'hf-bpe-metaspace',
): HfBpeMetaspaceTokenizer {
  return createHfBpeMetaspaceTokenizer(readFileSync(path, 'utf8'), name);
}

/**
 * Loads the Hf Bpe Metaspace Tokenizer From Url.
 */
export async function loadHfBpeMetaspaceTokenizerFromUrl(
  url: string,
  name = 'hf-bpe-metaspace',
): Promise<HfBpeMetaspaceTokenizer> {
  // The tokenizer is loaded once at child boot and lives for the whole process.
  // A single transient fetch failure (CDN blip, partial body) would otherwise
  // poison tokenization for the entire child lifetime — the live gemma4 garbage-
  // token symptom (2026-06-22). Retry with backoff + validate a non-trivial body.
  const ATTEMPTS = 5;
  let lastErr: unknown;
  for (let attempt = 1; attempt <= ATTEMPTS; attempt++) {
    try {
      const res = await fetch(url);
      if (!res.ok) {
        throw new Error(`HF tokenizer JSON fetch failed: ${res.status} ${res.statusText}`);
      }
      const text = await res.text();
      // Guard against a truncated/empty body silently producing a broken vocab.
      if (text.length < 1024 || !text.includes('"vocab"')) {
        throw new Error(`HF tokenizer JSON body looks invalid (len=${text.length})`);
      }
      return createHfBpeMetaspaceTokenizer(text, name);
    } catch (err) {
      lastErr = err;
      if (attempt < ATTEMPTS) {
        const backoffMs = 250 * 2 ** (attempt - 1);
        await new Promise((r) => setTimeout(r, backoffMs));
      }
    }
  }
  throw new Error(
    `HF tokenizer JSON fetch failed after ${ATTEMPTS} attempts (${url}): ${String(lastErr)}`,
  );
}
