// @a0n/gnosis/distributed-inference-host/src/homophonic-substitution-solver.ts
// homophonic substitution cipher solver. Two inner-loop strategies:
//
//   'hill-climb' (legacy): swap two symbols' letter assignments; accept
//     only if quadgram score improves. Greedy; gets stuck in local optima.
//     Pair-swap-only by design — this is the function the test suite
//     asserts on; reassignment lives in SA exclusively.
//
//   'simulated-annealing' (production-grade): two proposal kinds —
//     (a) 80% pair-swap two symbols' letter assignments, and
//     (b) 20% reassign one symbol to a random letter A-Z.
//     Accept improvements always; accept worsenings with probability
//     exp(delta / T) where T cools geometrically. The reassignment
//     mutation is the SA enhancement that defeats the homophone-group
//     plateau: pair-swap preserves the multiset of letters across the
//     mapping (5 symbols → 'E' stays 5 symbols → 'E' under any sequence
//     of swaps), so it can never adjust homophone group sizes. Z408's
//     optimal mapping needs ~6 symbols on 'E' (E is 12.7% of English);
//     reassignment is the only proposal that can grow or shrink those
//     groups. SA's downhill tolerance is what makes it tolerable as an
//     inner-loop move — most reassignments hurt the score.
//     This is what AZdecrypt uses to crack Z408-class ciphers in hours
//     instead of weeks.
//
//     Third SA enhancement: bigram-aware mutation bias. Uniform-random
//     symbol selection wastes iterations mutating already-correct
//     symbols. Instead, every SUSPICION_REFRESH_INTERVAL swaps we walk
//     the current plaintext, find positions where the (plaintext[i],
//     plaintext[i+1]) bigram has unusually-low English log-probability,
//     and bump a per-symbol "suspicion" counter for the cipher symbols
//     at those positions. We normalize by occurrence count and floor at
//     a small positive value (so every symbol still has nonzero pick
//     probability), then build a CDF for O(log n) weighted sampling.
//     The `i` in pair-swap and reassignment proposals is drawn from
//     this distribution; `j` in pair-swap stays uniform. The refresh
//     cadence (every 5000 swaps) amortizes the O(text_length) cost
//     against ~5000 SA iterations — negligible overhead — while
//     tracking the evolving decrypt: as suspicious symbols become
//     correct, the suspicion mass shifts to whatever's still wrong.
//
// FRF wrapping: N forks, each a full inner-loop session from a random
// initial mapping; fold by Buleyean rank on best-ever score.
// The Z408 cipher is solvable via SA + sufficient time/restarts; longer
// homophonic ciphers (Z340 with path transposition) need additional
// sieves layered in. Honest scope: structural solver, not a learned model.

import {
  quadgramScore,
  ENGLISH_MONOGRAM_LOG10,
  ENGLISH_BIGRAM_LOG10,
  ENGLISH_BIGRAM_FLOOR_LOG10,
} from './english-frequency';
import { buleyeanWeight } from '@a0n/buleyean-kernel';

/** How often (in swap counts) the SA loop recomputes per-symbol
 *  suspicion weights from the current plaintext. ~ciphertext.length
 *  ops per refresh, amortized over this many iterations. */
const SUSPICION_REFRESH_INTERVAL = 5000;

export interface HomophonicSolverArgs {
  /** Cipher text — any string; each unique character is treated as a
   *  distinct cipher symbol. Whitespace is preserved in the decrypted
   *  output but the symbol mapping is on non-whitespace characters. */
  ciphertext: string;
  /** Number of FRF forks (random-restart inner sessions). Default 8. */
  forks?: number;
  /** Max swaps per inner-loop session. Default 5000. */
  maxSwapsPerFork?: number;
  /** Stop a hill-climb after this many consecutive failed swaps.
   *  Ignored when strategy === 'simulated-annealing' (SA runs to T_end).
   *  Default 200. */
  stagnationLimit?: number;
  /** Random seed. Default 0xC0DE5EED. */
  seed?: number;
  /** Inner-loop strategy. Default 'hill-climb' (legacy / backwards-compat).
   *  Use 'simulated-annealing' for production cipher attacks — significantly
   *  better escape from local optima. */
  strategy?: 'hill-climb' | 'simulated-annealing';
  /** Simulated annealing tuning. Used only when strategy === 'simulated-annealing'. */
  annealing?: {
    /** Initial temperature. Default 5.0. Higher T at start = more
     *  worsening swaps accepted. For quadgram score deltas in [0, 2.0],
     *  T_start=5.0 makes ~80% of moderate (delta=-1) drops acceptable
     *  early on. */
    temperatureStart?: number;
    /** Final temperature. Default 0.01. T <= T_end = effectively
     *  hill-climb at the end. */
    temperatureEnd?: number;
    /** Geometric cooling rate per swap. Default 0.9999. Total iterations
     *  ≈ log(T_end / T_start) / log(coolingRate). With defaults that's
     *  ~62,000 iterations per fork. */
    coolingRate?: number;
  };
}

export interface HomophonicSolverCandidate {
  /** Quadgram score of the decrypted plaintext. */
  score: number;
  /** Decrypted plaintext (whitespace preserved). */
  plaintext: string;
  /** Symbol → letter mapping that produced this plaintext. */
  mapping: ReadonlyMap<string, string>;
  /** Number of swaps the hill-climb made. */
  swapCount: number;
  /** 0-based rank by descending score. */
  rank: number;
  /** buleyeanWeight(forks, rank). */
  buleyWeight: number;
}

export interface HomophonicSolverResult {
  /** Best-found plaintext (highest quadgram score). */
  plaintext: string;
  /** Best mapping. */
  mapping: ReadonlyMap<string, string>;
  /** Best score. */
  score: number;
  /** All forks ranked by score. */
  candidates: ReadonlyArray<HomophonicSolverCandidate>;
  /** Cipher symbol set (the unique non-whitespace characters in the input). */
  symbols: ReadonlyArray<string>;
}

const ALPHABET = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
// English letter frequency order, most-common first.
const ENGLISH_LETTERS_BY_FREQ = 'ETAOINSHRDLCUMWFGYPBVKJXQZ';

function lcg(seed: number) {
  let s = seed >>> 0;
  return () => {
    s = (s * 1103515245 + 12345) & 0x7fffffff;
    return s;
  };
}

function isWhitespace(c: string): boolean {
  return c === ' ' || c === '\t' || c === '\n' || c === '\r' || c === '\f' || c === '\v';
}

/** Walk ciphertext; for non-whitespace symbols look up the letter, else preserve. */
function decryptWithMapping(
  ciphertext: string,
  mapping: Map<string, string>,
): string {
  let out = '';
  for (let i = 0; i < ciphertext.length; i++) {
    const c = ciphertext[i];
    if (isWhitespace(c)) {
      out += c;
      continue;
    }
    const letter = mapping.get(c);
    out += letter ?? c;
  }
  return out;
}

/** Frequency-seeded initial mapping. Most-common symbol → 'E', etc. */
function frequencySeededMapping(
  symbols: ReadonlyArray<string>,
  ciphertext: string,
): Map<string, string> {
  const counts = new Map<string, number>();
  for (const s of symbols) counts.set(s, 0);
  for (let i = 0; i < ciphertext.length; i++) {
    const c = ciphertext[i];
    if (counts.has(c)) counts.set(c, (counts.get(c) ?? 0) + 1);
  }
  // Symbols sorted by descending count; ties broken by symbol string for determinism.
  const sorted = [...symbols].sort((a, b) => {
    const da = counts.get(a) ?? 0;
    const db = counts.get(b) ?? 0;
    if (db !== da) return db - da;
    return a < b ? -1 : a > b ? 1 : 0;
  });
  const mapping = new Map<string, string>();
  for (let i = 0; i < sorted.length; i++) {
    // Cycle through ETAOIN... when there are more symbols than 26.
    const letter = ENGLISH_LETTERS_BY_FREQ[i % 26];
    mapping.set(sorted[i], letter);
  }
  return mapping;
}

/** Fisher-Yates shuffle in place using LCG. */
function shuffleInPlace<T>(arr: T[], next: () => number): void {
  for (let i = arr.length - 1; i > 0; i--) {
    const j = next() % (i + 1);
    const tmp = arr[i];
    arr[i] = arr[j];
    arr[j] = tmp;
  }
}

/** Build a randomized initial mapping for a given fork. Symbols are still
 *  assigned letters cycling through ENGLISH_LETTERS_BY_FREQ, but the
 *  symbol order is a random permutation rather than the frequency order.
 *  Fork 0 is the deterministic frequency-seeded mapping (no shuffle). */
function randomInitialMapping(
  symbols: ReadonlyArray<string>,
  ciphertext: string,
  forkIndex: number,
  seed: number,
): Map<string, string> {
  if (forkIndex === 0) {
    return frequencySeededMapping(symbols, ciphertext);
  }
  const next = lcg((seed * 31 + forkIndex) >>> 0);
  const order = [...symbols];
  shuffleInPlace(order, next);
  const mapping = new Map<string, string>();
  for (let i = 0; i < order.length; i++) {
    mapping.set(order[i], ENGLISH_LETTERS_BY_FREQ[i % 26]);
  }
  return mapping;
}

interface InnerLoopOutcome {
  mapping: Map<string, string>;
  plaintext: string;
  score: number;
  swapCount: number;
}
type HillClimbOutcome = InnerLoopOutcome;

/** Hill-climb a mapping: random pairs of symbols swap their letter
 *  assignments. Keep swap if quadgram score improves. */
function hillClimb(
  ciphertext: string,
  symbols: ReadonlyArray<string>,
  initial: Map<string, string>,
  maxSwaps: number,
  stagnationLimit: number,
  next: () => number,
): HillClimbOutcome {
  const mapping = new Map(initial);
  let plaintext = decryptWithMapping(ciphertext, mapping);
  let score = quadgramScore(plaintext);
  let consecutiveFails = 0;
  let swapCount = 0;
  const n = symbols.length;
  if (n < 2) {
    return { mapping, plaintext, score, swapCount };
  }
  while (swapCount < maxSwaps && consecutiveFails < stagnationLimit) {
    const i = next() % n;
    let j = next() % n;
    if (j === i) j = (j + 1) % n;
    const sa = symbols[i];
    const sb = symbols[j];
    const la = mapping.get(sa)!;
    const lb = mapping.get(sb)!;
    if (la === lb) {
      // No-op swap; treat as a failed step but cheap.
      consecutiveFails++;
      swapCount++;
      continue;
    }
    mapping.set(sa, lb);
    mapping.set(sb, la);
    const candidatePlain = decryptWithMapping(ciphertext, mapping);
    const candidateScore = quadgramScore(candidatePlain);
    if (candidateScore > score) {
      plaintext = candidatePlain;
      score = candidateScore;
      consecutiveFails = 0;
    } else {
      // Revert.
      mapping.set(sa, la);
      mapping.set(sb, lb);
      consecutiveFails++;
    }
    swapCount++;
  }
  return { mapping, plaintext, score, swapCount };
}

/** Per-symbol suspicion weights from the current plaintext. For each
 *  position i, examine the bigram (plaintext[i], plaintext[i+1]); if
 *  its English log-probability is unusually low (below half the floor),
 *  bump suspicion for the cipher symbols at positions i and i+1.
 *  Normalize by occurrence count, floor at 0.01 so every symbol keeps
 *  nonzero pick probability. */
function computeSuspicionWeights(
  plaintext: string,
  ciphertext: string,
  symbols: ReadonlyArray<string>,
  symbolToIndex: Map<string, number>,
): number[] {
  const suspicion = new Array(symbols.length).fill(0);
  const counts = new Array(symbols.length).fill(0);
  // Half the floor: bigrams flagged "rare" are the long tail well
  // below the typical floor-clamped score. ENGLISH_BIGRAM_FLOOR_LOG10
  // is negative, so * 0.5 yields a less-extreme cutoff (e.g. -2.5).
  const RARE_THRESHOLD = ENGLISH_BIGRAM_FLOOR_LOG10 * 0.5;

  for (let i = 0; i < ciphertext.length - 1; i++) {
    const ci = ciphertext[i];
    const ci1 = ciphertext[i + 1];
    const idxA = symbolToIndex.get(ci);
    const idxB = symbolToIndex.get(ci1);
    if (idxA !== undefined) counts[idxA]++;
    if (idxB !== undefined) counts[idxB]++;

    const pi = plaintext[i];
    const pi1 = plaintext[i + 1];
    const bigram = pi + pi1;
    const logProb = ENGLISH_BIGRAM_LOG10.get(bigram) ?? ENGLISH_BIGRAM_FLOOR_LOG10;
    if (logProb < RARE_THRESHOLD) {
      if (idxA !== undefined) suspicion[idxA]++;
      if (idxB !== undefined) suspicion[idxB]++;
    }
  }

  const weights = new Array(symbols.length);
  for (let k = 0; k < symbols.length; k++) {
    const c = counts[k];
    weights[k] = c > 0 ? (suspicion[k] / c) + 0.01 : 0.01;
  }
  return weights;
}

/** Cumulative distribution function over `weights`. cdf[k] = sum of
 *  weights[0..k]; total mass is cdf[last]. */
function buildCDF(weights: number[]): number[] {
  const cdf = new Array(weights.length);
  let acc = 0;
  for (let k = 0; k < weights.length; k++) {
    acc += weights[k];
    cdf[k] = acc;
  }
  return cdf;
}

/** Sample an index from the CDF given uniform u in [0, 1). Binary
 *  search for the smallest k with cdf[k] >= u * total_mass. */
function sampleFromCDF(cdf: number[], u: number): number {
  const target = u * cdf[cdf.length - 1];
  let lo = 0;
  let hi = cdf.length - 1;
  while (lo < hi) {
    const mid = (lo + hi) >>> 1;
    if (cdf[mid] < target) lo = mid + 1;
    else hi = mid;
  }
  return lo;
}

/** Simulated annealing on the symbol-to-letter mapping. Same swap proposal
 *  as hill-climb; the difference is the acceptance rule:
 *    delta > 0:        accept (always)
 *    delta <= 0:       accept with probability exp(delta / T)
 *  T cools geometrically: T_{k+1} = T_k * coolingRate.
 *  Tracks best-ever (mapping, score) separately from current — final
 *  return is the best-ever, not the last accepted state.
 *  Symbol-selection bias: `i` is drawn from a bigram-suspicion CDF
 *  refreshed every SUSPICION_REFRESH_INTERVAL swaps. See header doc. */
function simulatedAnnealing(
  ciphertext: string,
  symbols: ReadonlyArray<string>,
  initial: Map<string, string>,
  maxSwaps: number,
  temperatureStart: number,
  temperatureEnd: number,
  coolingRate: number,
  next: () => number,
): InnerLoopOutcome {
  const mapping = new Map(initial);
  let plaintext = decryptWithMapping(ciphertext, mapping);
  let score = quadgramScore(plaintext);
  // Track best-ever: SA can wander into worse states but the final answer
  // is the best mapping the search ever visited.
  let bestMapping = new Map(mapping);
  let bestPlaintext = plaintext;
  let bestScore = score;
  let swapCount = 0;
  const n = symbols.length;
  if (n < 2) {
    return { mapping: bestMapping, plaintext: bestPlaintext, score: bestScore, swapCount };
  }
  // Build symbol→index lookup once for the bigram-suspicion sampler.
  const symbolToIndex = new Map<string, number>();
  for (let k = 0; k < symbols.length; k++) symbolToIndex.set(symbols[k], k);

  // Bigram-aware suspicion weights, refreshed periodically.
  let suspicionWeights = computeSuspicionWeights(
    plaintext,
    ciphertext,
    symbols,
    symbolToIndex,
  );
  let suspicionCDF = buildCDF(suspicionWeights);

  let T = temperatureStart;
  while (swapCount < maxSwaps && T > temperatureEnd) {
    // Refresh the suspicion CDF every SUSPICION_REFRESH_INTERVAL swaps.
    // The decrypt evolves; suspicious symbols at swap 0 are not the
    // suspicious symbols at swap 30000.
    if (swapCount > 0 && swapCount % SUSPICION_REFRESH_INTERVAL === 0) {
      suspicionWeights = computeSuspicionWeights(
        plaintext,
        ciphertext,
        symbols,
        symbolToIndex,
      );
      suspicionCDF = buildCDF(suspicionWeights);
    }

    // Two proposal kinds. 80% pair-swap (preserves the multiset of
    // letters across the mapping), 20% reassignment (one symbol → a
    // random letter A-Z; CAN change homophone group sizes — exactly
    // what pair-swap can't do, and exactly what the Z408 plateau at
    // score=-7.846 was diagnostic of).
    const proposalKind = (next() & 0x7fffffff) / 0x7fffffff;
    // Bigram-aware weighted pick for `i` — bias toward symbols that
    // participate in rare bigrams in the current plaintext. `j` (the
    // partner symbol in pair-swap, below) stays uniform.
    const uPick = (next() & 0x7fffffff) / 0x7fffffff;
    const i = sampleFromCDF(suspicionCDF, uPick);
    const sa = symbols[i];
    const la = mapping.get(sa)!;

    let newLetter: string;
    let revertLetter: string;
    let pairSb: string | undefined;
    let pairLb: string | undefined;

    if (proposalKind < 0.8) {
      // Pair-swap (existing behavior, gated by the 0.8 ratio).
      let j = next() % n;
      if (j === i) j = (j + 1) % n;
      pairSb = symbols[j];
      pairLb = mapping.get(pairSb)!;
      if (la === pairLb) {
        // No-op; tick clock.
        swapCount++;
        T *= coolingRate;
        continue;
      }
      mapping.set(sa, pairLb);
      mapping.set(pairSb, la);
      newLetter = pairLb;
      revertLetter = la;
    } else {
      // Reassignment: pick a random letter A-Z. May collide with
      // letters already used by other symbols (growing a homophone
      // group) or open up an unused letter (shrinking one).
      const letterIdx = next() % 26;
      newLetter = String.fromCharCode(65 + letterIdx);
      if (newLetter === la) {
        // No-op; tick clock.
        swapCount++;
        T *= coolingRate;
        continue;
      }
      mapping.set(sa, newLetter);
      revertLetter = la;
    }

    // Score the candidate.
    const candidatePlain = decryptWithMapping(ciphertext, mapping);
    const candidateScore = quadgramScore(candidatePlain);
    const delta = candidateScore - score;

    let accept = false;
    if (delta > 0) {
      accept = true;
    } else {
      const u = (next() & 0x7fffffff) / 0x7fffffff;
      if (u < Math.exp(delta / T)) accept = true;
    }

    if (accept) {
      plaintext = candidatePlain;
      score = candidateScore;
      if (score > bestScore) {
        bestScore = score;
        bestPlaintext = plaintext;
        bestMapping = new Map(mapping);
      }
    } else {
      // Revert. Branches by proposal type.
      if (pairSb !== undefined && pairLb !== undefined) {
        mapping.set(sa, revertLetter);
        mapping.set(pairSb, pairLb);
      } else {
        mapping.set(sa, revertLetter);
      }
    }

    swapCount++;
    T *= coolingRate;
  }
  return { mapping: bestMapping, plaintext: bestPlaintext, score: bestScore, swapCount };
}

/**
 * Handles the gnosis solve Homophonic Substitution workflow.
 */
export function solveHomophonicSubstitution(
  args: HomophonicSolverArgs,
): HomophonicSolverResult {
  const ciphertext = args.ciphertext;
  if (!ciphertext || ciphertext.length === 0) {
    throw new RangeError('ciphertext must be non-empty');
  }
  const forks = args.forks ?? 8;
  const maxSwapsPerFork = args.maxSwapsPerFork ?? 5000;
  const stagnationLimit = args.stagnationLimit ?? 200;
  const seed = (args.seed ?? 0xc0de5eed) >>> 0;
  if (!Number.isInteger(forks) || forks < 1) {
    throw new RangeError('forks must be an integer >= 1');
  }
  if (!Number.isInteger(maxSwapsPerFork) || maxSwapsPerFork < 100) {
    throw new RangeError('maxSwapsPerFork must be an integer >= 100');
  }
  if (!Number.isInteger(stagnationLimit) || stagnationLimit < 10) {
    throw new RangeError('stagnationLimit must be an integer >= 10');
  }

  // Reference monogram table; presence asserts the import resolves.
  if (ENGLISH_MONOGRAM_LOG10.length !== 26) {
    throw new Error('ENGLISH_MONOGRAM_LOG10 must have 26 entries');
  }

  // Extract unique non-whitespace symbols.
  const symbolSet = new Set<string>();
  for (let i = 0; i < ciphertext.length; i++) {
    const c = ciphertext[i];
    if (!isWhitespace(c)) symbolSet.add(c);
  }
  const symbols: ReadonlyArray<string> = [...symbolSet].sort();

  if (symbols.length === 0) {
    throw new RangeError('ciphertext must contain at least one non-whitespace symbol');
  }

  // FORK: run `forks` hill-climbs from random initial mappings.
  const rawCandidates: Array<{
    mapping: Map<string, string>;
    plaintext: string;
    score: number;
    swapCount: number;
  }> = [];

  const strategy = args.strategy ?? 'hill-climb';
  const annealing = args.annealing ?? {};
  const tStart = annealing.temperatureStart ?? 5.0;
  const tEnd = annealing.temperatureEnd ?? 0.01;
  const cool = annealing.coolingRate ?? 0.9999;
  if (strategy === 'simulated-annealing') {
    if (!Number.isFinite(tStart) || tStart <= 0) {
      throw new RangeError('annealing.temperatureStart must be > 0');
    }
    if (!Number.isFinite(tEnd) || tEnd <= 0 || tEnd >= tStart) {
      throw new RangeError(
        'annealing.temperatureEnd must satisfy 0 < tEnd < tStart',
      );
    }
    if (!Number.isFinite(cool) || cool <= 0 || cool >= 1) {
      throw new RangeError('annealing.coolingRate must be in (0, 1)');
    }
  }

  for (let f = 0; f < forks; f++) {
    const initial = randomInitialMapping(symbols, ciphertext, f, seed);
    const next = lcg((seed * 31 + f + 0x1234) >>> 0);
    const outcome = strategy === 'simulated-annealing'
      ? simulatedAnnealing(
          ciphertext,
          symbols,
          initial,
          maxSwapsPerFork,
          tStart,
          tEnd,
          cool,
          next,
        )
      : hillClimb(
          ciphertext,
          symbols,
          initial,
          maxSwapsPerFork,
          stagnationLimit,
          next,
        );
    rawCandidates.push(outcome);
  }

  // FOLD: rank by descending score, compute Buleyean weights.
  const sorted = [...rawCandidates].sort((a, b) => b.score - a.score);
  const candidates: HomophonicSolverCandidate[] = sorted.map((c, idx) => ({
    score: c.score,
    plaintext: c.plaintext,
    mapping: new Map(c.mapping),
    swapCount: c.swapCount,
    rank: idx,
    buleyWeight: buleyeanWeight(forks, idx),
  }));

  const best = candidates[0];
  return {
    plaintext: best.plaintext,
    mapping: best.mapping,
    score: best.score,
    candidates,
    symbols,
  };
}
