// @a0n/gnosis/distributed-inference-host/src/horn-observability-runtime.ts
import type { SpectralNoiseAssessment } from './spectral-noise-monitor';

export const HORN_OBSERVABILITY_RUNTIME_SCHEMA_VERSION =
  'horn_observability_runtime/v1' as const;

export const HORN_OBSERVABILITY_THEOREM_LINEAGE = [
  'Gnosis.HornObservabilityMetrics.matched_horn_escape_observability_schema',
  'Gnosis.HornObservabilityMetrics.canonical_witnesses_improve_observability_for_fixed_carrier',
  'Gnosis.HornLayeredReadiness.matched_horn_readiness_eq_carrier_readiness',
  'Gnosis.HornLayeredResidue.horn_escape_maps_to_unresolved_residue',
] as const;

export type HornObservabilityRuntimeSource =
  | 'attention_closure'
  | 'spectral_trace';

export interface HornObservabilityRuntimeInput {
  readonly boundaryAtResolution: number;
  readonly stableRows: number;
  readonly witnessScore?: number;
  readonly baselineWitnessScore?: number;
  readonly source?: HornObservabilityRuntimeSource;
  readonly observedUnresolvedResidue?: number;
}

export interface HornObservabilityRuntimeWire {
  readonly schemaVersion: typeof HORN_OBSERVABILITY_RUNTIME_SCHEMA_VERSION;
  readonly source: HornObservabilityRuntimeSource;
  readonly boundaryAtResolution: number;
  readonly stableRows: number;
  readonly unresolvedResidue: number;
  readonly observedUnresolvedResidue?: number;
  readonly residueMatchesObservation?: boolean;
  readonly absorbedInfo: number;
  readonly witnessScore: number;
  readonly baselineWitnessScore: number;
  readonly baselineObservabilityScore: number;
  readonly witnessedObservabilityScore: number;
  readonly witnessLift: number;
  readonly carrierReadiness: number;
  readonly carrierReadinessDeficit: number;
  readonly carrierReady: boolean;
  readonly partialObservation: boolean;
  readonly recoverableWithWitness: boolean;
  readonly theoremLineage: readonly string[];
}

function natural(value: number): number {
  if (!Number.isFinite(value)) return 0;
  return Math.max(0, Math.trunc(value));
}

/**
 * Computes the Horn Observability Runtime.
 */
export function computeHornObservabilityRuntime(
  input: HornObservabilityRuntimeInput,
): HornObservabilityRuntimeWire {
  const boundaryAtResolution = natural(input.boundaryAtResolution);
  const stableRows = natural(input.stableRows);
  const witnessScore = natural(input.witnessScore ?? 2);
  const baselineWitnessScore = natural(input.baselineWitnessScore ?? 0);
  const unresolvedResidue = Math.max(0, boundaryAtResolution - stableRows);
  const absorbedInfo = boundaryAtResolution - unresolvedResidue;
  const baselineObservabilityScore =
    absorbedInfo + baselineWitnessScore - unresolvedResidue;
  const witnessedObservabilityScore =
    absorbedInfo + witnessScore - unresolvedResidue;
  const carrierReadiness = 2 * stableRows - boundaryAtResolution;
  const carrierReadinessDeficit = Math.max(
    0,
    boundaryAtResolution - 2 * stableRows,
  );
  const observedUnresolvedResidue =
    input.observedUnresolvedResidue === undefined
      ? undefined
      : natural(input.observedUnresolvedResidue);

  return {
    schemaVersion: HORN_OBSERVABILITY_RUNTIME_SCHEMA_VERSION,
    source: input.source ?? 'spectral_trace',
    boundaryAtResolution,
    stableRows,
    unresolvedResidue,
    observedUnresolvedResidue,
    residueMatchesObservation:
      observedUnresolvedResidue === undefined
        ? undefined
        : observedUnresolvedResidue === unresolvedResidue,
    absorbedInfo,
    witnessScore,
    baselineWitnessScore,
    baselineObservabilityScore,
    witnessedObservabilityScore,
    witnessLift: witnessedObservabilityScore - baselineObservabilityScore,
    carrierReadiness,
    carrierReadinessDeficit,
    carrierReady: carrierReadiness >= 0,
    partialObservation: unresolvedResidue > 0,
    recoverableWithWitness: unresolvedResidue > 0 && witnessScore > 0,
    theoremLineage: HORN_OBSERVABILITY_THEOREM_LINEAGE,
  };
}

/**
 * Computes the Horn Observability From Spectral Assessment.
 */
export function computeHornObservabilityFromSpectralAssessment(
  spectral: SpectralNoiseAssessment,
): HornObservabilityRuntimeWire {
  const ac = spectral.attentionClosure;
  if (ac !== undefined) {
    return computeHornObservabilityRuntime({
      boundaryAtResolution: ac.observed.sourceResolution,
      stableRows: ac.observed.observerResolution,
      observedUnresolvedResidue: ac.observed.unresolvedResidue,
      source: 'attention_closure',
    });
  }

  return computeHornObservabilityRuntime({
    boundaryAtResolution: spectral.trace.requestEntropy,
    stableRows: spectral.trace.boundaryLimit,
    source: 'spectral_trace',
  });
}
