// @a0n/gnosis/distributed-inference-host/src/image-noise-estimate.ts
// Adaptive noise estimator for the image-carrier deblur pipeline.
// Implements Immerkaer 1996 -- convolves the image with a double-Laplacian
// kernel, then aggregates absolute response over the interior. Robust to
// image content because smooth gradients pass through the kernel as ~0
// while noise responds at full amplitude. Used by deblur-any-image when
// the caller passes `noiseFloor: 'auto'`.

export interface EstimateNoiseStdArgs {
  plane: Float32Array; // length width * height
  width: number;
  height: number;
}

function validate(args: EstimateNoiseStdArgs): void {
  const { plane, width, height } = args;
  if (width < 3 || height < 3) {
    throw new RangeError(
      `estimateNoiseStd: width and height must each be >= 3, got w=${width} h=${height}`,
    );
  }
  if (plane.length !== width * height) {
    throw new RangeError(
      `estimateNoiseStd: plane.length=${plane.length} does not match width*height=${width * height}`,
    );
  }
}

/** Estimated noise standard deviation in the same units as `plane`
 *  (i.e. for a [0,1] linear plane, sigma_hat is also in [0,1]).
 *  Throws RangeError on shape mismatch or w/h < 3. */
export function estimateNoiseStd(args: EstimateNoiseStdArgs): number {
  validate(args);
  const { plane, width: w, height: h } = args;

  // Double-Laplacian kernel (Immerkaer 1996):
  //   1  -2   1
  //  -2   4  -2
  //   1  -2   1
  // Apply at interior pixels only (1 <= x <= w-2, 1 <= y <= h-2).
  // Fold kernel inline -- no intermediate allocation.
  let sumAbs = 0;
  for (let y = 1; y <= h - 2; y++) {
    const rowM1 = (y - 1) * w;
    const row0 = y * w;
    const rowP1 = (y + 1) * w;
    for (let x = 1; x <= w - 2; x++) {
      const tl = plane[rowM1 + (x - 1)];
      const tc = plane[rowM1 + x];
      const tr = plane[rowM1 + (x + 1)];
      const ml = plane[row0 + (x - 1)];
      const mc = plane[row0 + x];
      const mr = plane[row0 + (x + 1)];
      const bl = plane[rowP1 + (x - 1)];
      const bc = plane[rowP1 + x];
      const br = plane[rowP1 + (x + 1)];

      const L =
        tl - 2 * tc + tr +
        -2 * ml + 4 * mc - 2 * mr +
        bl - 2 * bc + br;

      sumAbs += L < 0 ? -L : L;
    }
  }

  const interiorW = w - 2;
  const interiorH = h - 2;
  const norm = Math.sqrt(Math.PI / 2) / (6 * interiorW * interiorH);
  return sumAbs * norm;
}

/** Wiener-friendly noise floor: sigma_hat^2 scaled into the same units as
 *  |H_hat|^2 for a sum-1-normalized PSF on a unit-range plane. Floor at
 *  1e-5 so flat synthetic inputs don't produce a zero floor and divide
 *  by zero. */
export function estimateNoiseFloor(args: EstimateNoiseStdArgs): number {
  const sigma = estimateNoiseStd(args);
  return Math.max(1e-5, sigma * sigma);
}
