// @a0n/gnosis/distributed-inference-host/src/image-deblur.ts
// image-deblur.ts — canonical reference deblur for the image-carrier smoke.
// Single-shot Wiener filter in the frequency domain with a known PSF.
// Iterative variants (Richardson–Lucy, FRF over PSF candidates) are deferred to Phase 2.
//
// noiseFloor is the regularization parameter that prevents division by zeros in
// |Ĥ|² (spectral nulls of the blur kernel). Lower = sharper but ringier; higher
// = smoother but blurrier residual. Default 1e-3 is tuned to clear ΔPSNR ≥ 6 dB
// on the smoke fixture.
//
// Recovered output is NOT clamped to [0, 1]: regularization edge effects can
// push values slightly outside that range, and the caller measures PSNR against
// the original (which lives in [0, 1]). Clamping here would mask Wiener bias.

import { fft2, ifft2 } from './fft2';
import { padAndCenterPSF } from './image-psf';

export interface MeshResidualDeblurConfig {
  overpotential: number;
  admission_floor: number;
  saturation_limit: number;
  depth_confidence_blend: number;
}

export interface DeblurCertificate {
  schema_version: string;
  width: number;
  height: number;
  overpotential: number;
  admission_floor: number;
  saturation_limit: number;
  admitted_pixels: number;
  mean_admission: number;
  mean_abs_residual_before: number;
  mean_abs_residual_after: number;
  max_channel_delta: number;
}

/**
 * Computes the Deblur Certificate.
 */
export function computeDeblurCertificate(
  before: Float32Array,
  after: Float32Array,
  w: number,
  h: number,
  channels: number,
): DeblurCertificate {
  let max_delta = 0;
  let admitted = 0;
  let res_after = 0;
  const pix = w * h;

  for (let i = 0; i < pix; i++) {
    let pixel_admitted = false;
    for (let c = 0; c < channels; c++) {
      const idx = i * channels + c;
      const src = before[idx]!;
      const dst = after[idx]!;
      const delta = Math.abs(dst - src);
      if (delta > max_delta) max_delta = delta;
      if (delta > 1e-6) pixel_admitted = true;
      res_after += delta;
    }
    if (pixel_admitted) admitted++;
  }

  const denom = pix * channels || 1;

  return {
    schema_version: 'mesh_residual_deblur/v1',
    width: w,
    height: h,
    overpotential: 0,
    admission_floor: 0,
    saturation_limit: 0,
    admitted_pixels: admitted,
    mean_admission: admitted / (pix || 1),
    mean_abs_residual_before: 0,
    mean_abs_residual_after: res_after / denom,
    max_channel_delta: max_delta,
  };
}

export interface WienerArgs {
  plane: Float32Array;        // length w*h
  w: number;                  // power of 2
  h: number;                  // power of 2
  psf: Float32Array;          // length psfW*psfH
  psfW: number;               // positive odd integer
  psfH: number;               // positive odd integer
  noiseFloor?: number;        // default 1e-3
}

const DEFAULT_NOISE_FLOOR = 1e-3;

function isPow2(n: number): boolean {
  return Number.isInteger(n) && n > 0 && (n & (n - 1)) === 0;
}

function isPositiveOdd(n: number): boolean {
  return Number.isInteger(n) && n > 0 && (n & 1) === 1;
}

/**
 * Handles the gnosis wiener Deblur workflow.
 */
export function wienerDeblur(args: WienerArgs): Float32Array {
  const { plane, w, h, psf, psfW, psfH } = args;
  const noiseFloor = args.noiseFloor ?? DEFAULT_NOISE_FLOOR;

  // Validate plane dimensions.
  if (!isPow2(w)) {
    throw new RangeError(`wienerDeblur: w must be a positive power of 2 (got ${w})`);
  }
  if (!isPow2(h)) {
    throw new RangeError(`wienerDeblur: h must be a positive power of 2 (got ${h})`);
  }
  if (plane.length !== w * h) {
    throw new RangeError(
      `wienerDeblur: plane.length (${plane.length}) must equal w*h (${w * h})`,
    );
  }

  // Validate PSF dimensions.
  if (!isPositiveOdd(psfW)) {
    throw new RangeError(`wienerDeblur: psfW must be a positive odd integer (got ${psfW})`);
  }
  if (!isPositiveOdd(psfH)) {
    throw new RangeError(`wienerDeblur: psfH must be a positive odd integer (got ${psfH})`);
  }
  if (psf.length !== psfW * psfH) {
    throw new RangeError(
      `wienerDeblur: psf.length (${psf.length}) must equal psfW*psfH (${psfW * psfH})`,
    );
  }
  if (psfW > w || psfH > h) {
    throw new RangeError(
      `wienerDeblur: PSF (${psfW}x${psfH}) must fit within plane (${w}x${h})`,
    );
  }

  // Validate noise floor.
  if (!Number.isFinite(noiseFloor) || noiseFloor <= 0) {
    throw new RangeError(
      `wienerDeblur: noiseFloor must be a positive finite number (got ${noiseFloor})`,
    );
  }

  // 1. Pad and center PSF to plane dimensions (DC at origin per padAndCenterPSF contract).
  const padded = padAndCenterPSF(psf, psfW, psfH, w, h);

  // 2. Forward FFTs of degraded image and PSF.
  const Y = fft2(plane, w, h);
  const H = fft2(padded, w, h);

  // 3. Build X̂ = (conj(Ĥ) / (|Ĥ|² + N)) · Ŷ in the frequency domain.
  const n = w * h;
  const xr = new Float32Array(n);
  const xi = new Float32Array(n);

  for (let i = 0; i < n; i++) {
    const hr = H.re[i];
    const hi = H.im[i];
    const yr = Y.re[i];
    const yi = Y.im[i];

    const magSq = hr * hr + hi * hi;
    const denom = magSq + noiseFloor;

    // W = conj(H) / denom = (hr / denom, -hi / denom)
    const wr = hr / denom;
    const wi = -hi / denom;

    // X = W * Y (complex multiply)
    xr[i] = wr * yr - wi * yi;
    xi[i] = wr * yi + wi * yr;
  }

  // 4. Inverse FFT — caller takes real part directly (ifft2 returns real component).
  return ifft2({ re: xr, im: xi, w, h });
}

/**
 * Handles the gnosis box Blur Rgb3 workflow.
 */
export function boxBlurRgb3(rgb: Float32Array, width: number, height: number): Float32Array {
  const pix = width * height;
  const out = new Float32Array(pix * 3);
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      let sumR = 0, sumG = 0, sumB = 0;
      let count = 0;
      for (let dy = -1; dy <= 1; dy++) {
        for (let dx = -1; dx <= 1; dx++) {
          const xx = Math.max(0, Math.min(width - 1, x + dx));
          const yy = Math.max(0, Math.min(height - 1, y + dy));
          const src = (yy * width + xx) * 3;
          sumR += Math.max(0, Math.min(1, rgb[src]!));
          sumG += Math.max(0, Math.min(1, rgb[src + 1]!));
          sumB += Math.max(0, Math.min(1, rgb[src + 2]!));
          count += 1;
        }
      }
      const dst = (y * width + x) * 3;
      out[dst] = sumR / count;
      out[dst + 1] = sumG / count;
      out[dst + 2] = sumB / count;
    }
  }
  return out;
}

export interface MeshResidualDeblurResult {
  rgb: Float32Array;
  admission: Float32Array;
  certificate: DeblurCertificate;
}

/**
 * Handles the gnosis mesh Residual Deblur workflow.
 */
export function meshResidualDeblur(
  rgb: Float32Array,
  width: number,
  height: number,
  topologyAlpha: Float32Array,
  depthConfidence: Float32Array,
  config: MeshResidualDeblurConfig,
): MeshResidualDeblurResult {
  const pix = width * height;
  const need = pix * 3;
  if (rgb.length !== need || topologyAlpha.length !== pix || depthConfidence.length !== pix) {
    throw new RangeError('meshResidualDeblur: invalid buffer sizes');
  }

  const blurred = boxBlurRgb3(rgb, width, height);
  const floor = Math.max(0, Math.min(1, config.admission_floor));
  const eta = Math.max(0, Math.min(2, config.overpotential));
  const sat = Math.max(0, Math.min(0.5, config.saturation_limit));
  const depthBlend = Math.max(0, Math.min(1, config.depth_confidence_blend));

  const out = new Float32Array(need);
  const admission = new Float32Array(pix);
  
  let admitted = 0;
  let admissionSum = 0;
  let residualBeforeSum = 0;
  let residualAfterSum = 0;
  let maxDelta = 0;

  for (let i = 0; i < pix; i++) {
    const topo = Math.max(0, Math.min(1, topologyAlpha[i]!));
    const depth = Math.max(0, Math.min(1, depthConfidence[i]!));
    const rawGate = topo * (1.0 - depthBlend) + depth * depthBlend;
    
    let gate = 0;
    if (rawGate > floor) {
      gate = Math.max(0, Math.min(1, (rawGate - floor) / Math.max(1e-6, 1.0 - floor)));
    }
    
    admission[i] = gate;
    admissionSum += gate;
    if (gate > 0) admitted++;

    const base = i * 3;
    for (let c = 0; c < 3; c++) {
      const src = Math.max(0, Math.min(1, rgb[base + c]!));
      const blur = Math.max(0, Math.min(1, blurred[base + c]!));
      const residual = src - blur;
      const lift = Math.max(-sat, Math.min(sat, eta * gate * residual));
      const dst = Math.max(0, Math.min(1, src + lift));
      
      out[base + c] = dst;
      residualBeforeSum += Math.abs(residual);
      residualAfterSum += Math.abs(dst - blur);
      maxDelta = Math.max(maxDelta, Math.abs(dst - src));
    }
  }

  const denom = need || 1;

  return {
    rgb: out,
    admission,
    certificate: {
      schema_version: 'mesh_residual_deblur/v1',
      width,
      height,
      overpotential: eta,
      admission_floor: floor,
      saturation_limit: sat,
      admitted_pixels: admitted,
      mean_admission: admissionSum / (pix || 1),
      mean_abs_residual_before: residualBeforeSum / denom,
      mean_abs_residual_after: residualAfterSum / denom,
      max_channel_delta: maxDelta,
    }
  };
}
