// @a0n/gnosis/distributed-inference-host/src/image-psf.ts
// image-psf.ts
// Forward-blur model used to manufacture the smoke's blurred input AND the
// kernel-shaping helper consumed by the Wiener deblur in `image-deblur.ts`.
// `applyPSF` is FFT-circular convolution (NOT spatial-with-edge-replication):
// the deblur is also FFT-circular, and the two MUST share the same convolution
// model so PSNR(recovered, original) measures the deblur math, not edge artifacts.
// Consumer: ./image-deblur.ts.

import { fft2, ifft2, type Complex2D } from './fft2';

/**
 * Recommended PSF kernel side length for a given σ. Returns the smallest
 * positive odd integer that covers ±3σ on each axis (99.7% of Gaussian
 * energy). Floor of 11 to keep small-σ kernels well-conditioned.
 *
 * Rationale: a Gaussian truncated below 6σ leaves visible energy in the
 * tails. Wiener deconvolution divides by |Ĥ|², so any ripple in Ĥ caused
 * by tail truncation gets amplified into spatial banding ("Gibbs / Doppler"
 * stripes). Scaling the kernel with σ keeps the truncation residual below
 * the noise floor.
 */
export function recommendPsfSize(sigma: number): number {
  if (!Number.isFinite(sigma) || sigma <= 0) {
    throw new RangeError(`recommendPsfSize: sigma must be > 0, got ${sigma}`);
  }
  const span = 2 * Math.ceil(3 * sigma) + 1; // already odd
  return span < 11 ? 11 : span;
}

/**
 * Build a centered 2D Gaussian kernel of side length `size`, summed to 1.0.
 *
 * @param sigma Standard deviation (must be > 0).
 * @param size  Side length in pixels (must be a positive odd integer so the
 *              kernel has a single pixel at its geometric center).
 * @returns     Length `size*size` row-major Float32Array, normalized.
 */
export function gaussianPSF(sigma: number, size: number): Float32Array {
  if (!Number.isFinite(sigma) || sigma <= 0) {
    throw new RangeError(`gaussianPSF: sigma must be > 0, got ${sigma}`);
  }
  if (!Number.isInteger(size) || size <= 0 || (size & 1) === 0) {
    throw new RangeError(`gaussianPSF: size must be a positive odd integer, got ${size}`);
  }

  const out = new Float32Array(size * size);
  const c = (size - 1) / 2;
  const twoSigma2 = 2 * sigma * sigma;

  let sum = 0;
  for (let y = 0; y < size; y++) {
    const dy = y - c;
    for (let x = 0; x < size; x++) {
      const dx = x - c;
      const v = Math.exp(-(dx * dx + dy * dy) / twoSigma2);
      out[y * size + x] = v;
      sum += v;
    }
  }

  if (!(sum > 0) || !Number.isFinite(sum)) {
    throw new RangeError(`gaussianPSF: degenerate kernel sum=${sum} (sigma=${sigma}, size=${size})`);
  }

  const inv = 1 / sum;
  for (let i = 0; i < out.length; i++) out[i] *= inv;
  return out;
}

/**
 * Zero-pad a centered PSF to (`w`, `h`) and wrap the PSF center to padded
 * position (0, 0). This is the standard FFT-convolution PSF placement —
 * circular convolution of an image with this padded kernel matches a
 * centered spatial convolution by the original PSF.
 *
 * @param psf   Source PSF, row-major, length psfW*psfH.
 * @param psfW  Source PSF width  (positive odd integer).
 * @param psfH  Source PSF height (positive odd integer).
 * @param w     Target padded width  (power of 2, >= psfW).
 * @param h     Target padded height (power of 2, >= psfH).
 */
export function padAndCenterPSF(
  psf: Float32Array,
  psfW: number,
  psfH: number,
  w: number,
  h: number,
): Float32Array {
  if (!Number.isInteger(psfW) || psfW <= 0 || (psfW & 1) === 0) {
    throw new RangeError(`padAndCenterPSF: psfW must be a positive odd integer, got ${psfW}`);
  }
  if (!Number.isInteger(psfH) || psfH <= 0 || (psfH & 1) === 0) {
    throw new RangeError(`padAndCenterPSF: psfH must be a positive odd integer, got ${psfH}`);
  }
  if (psf.length !== psfW * psfH) {
    throw new RangeError(`padAndCenterPSF: psf.length=${psf.length} != psfW*psfH=${psfW * psfH}`);
  }
  if (!Number.isInteger(w) || w <= 0 || (w & (w - 1)) !== 0) {
    throw new RangeError(`padAndCenterPSF: w must be a positive power of 2, got ${w}`);
  }
  if (!Number.isInteger(h) || h <= 0 || (h & (h - 1)) !== 0) {
    throw new RangeError(`padAndCenterPSF: h must be a positive power of 2, got ${h}`);
  }
  if (w < psfW || h < psfH) {
    throw new RangeError(`padAndCenterPSF: (w,h)=(${w},${h}) must be >= (psfW,psfH)=(${psfW},${psfH})`);
  }

  const padded = new Float32Array(w * h);
  const cx = psfW >> 1;
  const cy = psfH >> 1;

  for (let py = 0; py < psfH; py++) {
    const dy = py - cy;
    const ty = ((dy % h) + h) % h;
    for (let px = 0; px < psfW; px++) {
      const dx = px - cx;
      const tx = ((dx % w) + w) % w;
      padded[ty * w + tx] = psf[py * psfW + px];
    }
  }

  return padded;
}

export interface ApplyPSFArgs {
  plane: Float32Array;
  w: number;
  h: number;
  psf: Float32Array;
  psfW: number;
  psfH: number;
}

/**
 * FFT-based circular convolution of `plane` with `psf`. Returns the blurred
 * plane (same dimensions as input). This is the forward model the deblur
 * inverts; both share the FFT-circular boundary so PSNR reflects the math.
 */
export function applyPSF(args: ApplyPSFArgs): Float32Array {
  const { plane, w, h, psf, psfW, psfH } = args;

  if (!Number.isInteger(w) || w <= 0 || (w & (w - 1)) !== 0) {
    throw new RangeError(`applyPSF: w must be a positive power of 2, got ${w}`);
  }
  if (!Number.isInteger(h) || h <= 0 || (h & (h - 1)) !== 0) {
    throw new RangeError(`applyPSF: h must be a positive power of 2, got ${h}`);
  }
  if (plane.length !== w * h) {
    throw new RangeError(`applyPSF: plane.length=${plane.length} != w*h=${w * h}`);
  }

  const padded = padAndCenterPSF(psf, psfW, psfH, w, h);
  const Y: Complex2D = fft2(plane, w, h);
  const H: Complex2D = fft2(padded, w, h);

  const n = w * h;
  const zRe = new Float32Array(n);
  const zIm = new Float32Array(n);
  for (let i = 0; i < n; i++) {
    const yr = Y.re[i];
    const yi = Y.im[i];
    const hr = H.re[i];
    const hi = H.im[i];
    zRe[i] = yr * hr - yi * hi;
    zIm[i] = yr * hi + yi * hr;
  }

  const Z: Complex2D = { re: zRe, im: zIm, w, h };
  return ifft2(Z);
}
