// @a0n/gnosis/distributed-inference-host/src/image-fft-frf.ts
// Row-/column-tiled 2D FFT via the FRF primitive. Output bit-equivalent to
// fft2.ts on tiles=1; for tiles>=2 each FORK computes a contiguous block of
// rows (forward) or columns (inverse) of the 1D FFT pass. The fold
// concatenates tile outputs in input order. Today the inner 1D FFT is the
// same pure-JS Cooley-Tukey from fft2; this file is the structural slot the
// WASM SIMD plan plugs into. `tiles` is currently advisory: the JS reference
// path delegates to fft2/ifft2 inside a single FRF fork so the audit trail
// (report.fulfilledCount, etc.) stays consistent for callers.

import { fft2, ifft2, isPowerOfTwo, type Complex2D } from './fft2';
import { forkRaceFold, type ForkTask } from './frf';

export interface FFT2FRFArgs {
  /** Real input plane, length width*height. */
  real: Float32Array;
  width: number;
  height: number;
  /** Number of row tiles for the parallel structure. Default 1 (delegate to
   *  fft2 directly with no tiling — the bit-exact reference path). */
  tiles?: number;
  concurrency?: number;
}

export interface IFFT2FRFArgs {
  spectrum: Complex2D;
  tiles?: number;
  concurrency?: number;
}

function validateTiles(tiles: number | undefined): number {
  if (tiles === undefined) return 1;
  if (!Number.isInteger(tiles) || tiles <= 0) {
    throw new RangeError(
      `image-fft-frf: tiles must be a positive integer; got ${tiles}`,
    );
  }
  return tiles;
}

function validateConcurrency(concurrency: number | undefined): void {
  if (concurrency === undefined) return;
  if (!Number.isInteger(concurrency) || concurrency <= 0) {
    throw new RangeError(
      `image-fft-frf: concurrency must be a positive integer; got ${concurrency}`,
    );
  }
}

/**
 * Forward 2D FFT structured as a single-fork FRF over fft2. Bit-exact to
 * fft2.ts on the same input.
 *
 * `tiles` is advisory: it locks the FRF-shaped contract for callers (the
 * Rust SIMD implementation will honour it) but the JS reference always
 * runs the full plane in one fork. We deliberately do not fake row-tiled
 * parallelism on a single thread — Promise.all over slices buys nothing
 * and would only mask cost-algebra audits.
 */
export async function fft2FRF(args: FFT2FRFArgs): Promise<Complex2D> {
  const { real, width, height, concurrency } = args;
  const tiles = validateTiles(args.tiles);
  validateConcurrency(concurrency);

  if (!isPowerOfTwo(width)) {
    throw new RangeError(
      `image-fft-frf: width must be a positive power of two; got ${width}`,
    );
  }
  if (!isPowerOfTwo(height)) {
    throw new RangeError(
      `image-fft-frf: height must be a positive power of two; got ${height}`,
    );
  }
  if (real.length !== width * height) {
    throw new RangeError(
      `image-fft-frf: real.length ${real.length} != width*height ${width * height}`,
    );
  }
  // tiles is currently advisory; we record it via the fork count. Phase B
  // (Rust SIMD) will drive real per-row tiling.
  void tiles;

  const fork: ForkTask<Complex2D> = async () => fft2(real, width, height);
  const { output } = await forkRaceFold<Complex2D, Complex2D>({
    forks: [fork],
    fold: (survivors) => survivors[0]!,
    concurrency,
  });
  return output;
}

/**
 * Inverse 2D FFT structured as a single-fork FRF over ifft2. Bit-exact to
 * fft2.ts on the same input.
 */
export async function ifft2FRF(args: IFFT2FRFArgs): Promise<Float32Array> {
  const { spectrum, concurrency } = args;
  const tiles = validateTiles(args.tiles);
  validateConcurrency(concurrency);

  if (!spectrum || !(spectrum.re instanceof Float32Array) || !(spectrum.im instanceof Float32Array)) {
    throw new RangeError('image-fft-frf: spectrum must have re/im Float32Arrays');
  }
  const { w, h } = spectrum;
  if (!isPowerOfTwo(w)) {
    throw new RangeError(
      `image-fft-frf: spectrum.w must be a positive power of two; got ${w}`,
    );
  }
  if (!isPowerOfTwo(h)) {
    throw new RangeError(
      `image-fft-frf: spectrum.h must be a positive power of two; got ${h}`,
    );
  }
  if (spectrum.re.length !== w * h || spectrum.im.length !== w * h) {
    throw new RangeError(
      `image-fft-frf: spectrum re/im length must equal w*h ${w * h}`,
    );
  }
  void tiles;

  const fork: ForkTask<Float32Array> = async () => ifft2(spectrum);
  const { output } = await forkRaceFold<Float32Array, Float32Array>({
    forks: [fork],
    fold: (survivors) => survivors[0]!,
    concurrency,
  });
  return output;
}
