// @a0n/gnosis/distributed-inference-host/src/image-carrier-deblur.ts
// Phase 1 orchestrator for image-carrier deblur.
//
// Wires `ImageCarrierPipeline.analyze` (carrier witness) → `wienerDeblur`
// (single-shot Wiener) → `classifyDeblurStep` (host-side Bule-face label)
// → `composeWithFaceUnit` (the canonical +1 step on a BuleyProjection).
//
// Phase 1 produces ONE deblur step (Wiener is single-shot), so `perStep`
// has length 1 and `iterations === 1`. The API is iteration-shaped so
// Phase 2 can swap in iterative deblur (Richardson–Lucy or fork/race/fold
// over PSF candidates) without changing callers.
//
// Loud-fail policy: if the composed BuleyProjection.score exceeds the
// tower ceiling (default Hexon=6), this throws — silent score drift is
// the same class of bug as silent ledger zero-fill (see
// feedback_silent_catch_anti_pattern in memory).

import {
  ImageCarrierPipeline,
  type ImageCarrierAnalysis,
  type ImageCarrierInput,
  type ImageCarrierPipelineConfig,
} from './image-carrier-pipeline';
import { wienerDeblur } from './image-deblur';
import {
  classifyDeblurStep,
  type DeblurStepClassification,
} from './image-spectrum-classify';
import { composeWithFaceUnit } from './cost-algebra';
import {
  TOWER_CEILINGS,
  vacuumBuleyProjection,
  type BuleyFace,
  type BuleyProjection,
} from './spectral-noise-monitor';

export interface DeblurImageCarrierArgs {
  /** Single-channel luminance image in [0,1] (or byte range with sampleRange='byte'). */
  input: ImageCarrierInput;
  psf: Float32Array;
  psfW: number;
  psfH: number;
  /** Wiener regularization. Default 1e-3. */
  noiseFloor?: number;
  /** Tower ceiling on `carrierBuley.score`. Default `TOWER_CEILINGS.hexon` (6). */
  ceiling?: number;
  /** Forwarded to `ImageCarrierPipeline`. Default lets the pipeline choose. */
  pipelineConfig?: ImageCarrierPipelineConfig;
}

export interface DeblurStepReport {
  face: BuleyFace;
  reason: DeblurStepClassification['reason'];
  lowBandDelta: number;
  highBandDelta: number;
  totalDelta: number;
}

export interface DeblurImageCarrierResult {
  /** Recovered luminance plane, length width*height. NOT clamped to [0,1]. */
  recovered: Float32Array;
  /** Carrier witness from `ImageCarrierPipeline.analyze` over the BLURRED input. */
  analysis: ImageCarrierAnalysis;
  /** BuleyProjection composed from vacuum via one `composeWithFaceUnit` per step. */
  carrierBuley: BuleyProjection;
  /** One entry per deblur step; length === iterations. Length 1 in Phase 1. */
  perStep: ReadonlyArray<DeblurStepReport>;
  iterations: number;
}

/**
 * Handles the gnosis deblur Image Carrier workflow.
 */
export function deblurImageCarrier(args: DeblurImageCarrierArgs): DeblurImageCarrierResult {
  const { input, psf, psfW, psfH, noiseFloor } = args;
  const ceiling = args.ceiling ?? TOWER_CEILINGS.hexon;

  const channels = input.channels ?? 1;
  if (channels !== 1) {
    throw new RangeError(
      `deblurImageCarrier: Phase 1 supports only single-channel luminance (got channels=${channels})`,
    );
  }

  const w = input.width;
  const h = input.height;
  const expected = w * h;
  if (input.data.length !== expected) {
    throw new RangeError(
      `deblurImageCarrier: input.data.length=${input.data.length} != width*height=${expected}`,
    );
  }

  const plane = toUnitFloat32Plane(input, expected);

  const pipeline = new ImageCarrierPipeline(args.pipelineConfig ?? {});
  const analysis = pipeline.analyze(input);

  const recovered = wienerDeblur({ plane, w, h, psf, psfW, psfH, noiseFloor });

  const classification = classifyDeblurStep({ before: plane, after: recovered, w, h });
  const carrierBuley = composeWithFaceUnit(vacuumBuleyProjection, classification.face);

  if (carrierBuley.score > ceiling) {
    throw new Error(
      `deblurImageCarrier: carrierBuley.score=${carrierBuley.score} exceeds ceiling=${ceiling} ` +
        `(face=${classification.face}, reason=${classification.reason}). Tower contract violated.`,
    );
  }

  return {
    recovered,
    analysis,
    carrierBuley,
    perStep: [
      {
        face: classification.face,
        reason: classification.reason,
        lowBandDelta: classification.lowBandDelta,
        highBandDelta: classification.highBandDelta,
        totalDelta: classification.totalDelta,
      },
    ],
    iterations: 1,
  };
}

function toUnitFloat32Plane(input: ImageCarrierInput, expected: number): Float32Array {
  if (input.data instanceof Float32Array && input.sampleRange !== 'byte' && input.data.length === expected) {
    return input.data;
  }

  const out = new Float32Array(expected);
  const sr = input.sampleRange;

  if (sr === 'byte') {
    for (let i = 0; i < expected; i++) out[i] = (input.data[i] ?? 0) / 255;
    return out;
  }
  if (sr === 'unit') {
    for (let i = 0; i < expected; i++) out[i] = input.data[i] ?? 0;
    return out;
  }

  // Auto-detect: matches `pixelLuminance` in image-carrier-pipeline.ts — if
  // any sample exceeds 1 we assume byte range.
  let maxVal = 0;
  for (let i = 0; i < expected; i++) {
    const v = input.data[i] ?? 0;
    if (v > maxVal) maxVal = v;
  }
  const inv = maxVal > 1 ? 1 / 255 : 1;
  for (let i = 0; i < expected; i++) out[i] = (input.data[i] ?? 0) * inv;
  return out;
}
