// @a0n/gnosis/distributed-inference-host/src/image-deblur-rl.ts
// image-deblur-rl.ts — Richardson–Lucy iterative deconvolution.
// Switchable alternative to wienerDeblur; opt in via deblur-any-image's
// method:'richardson-lucy'. Naturally non-negative, less ringing on
// textured regions than Wiener, but iterations × FFT cost — ~5-10× slower
// than single-shot Wiener for a typical 20-iter run. Symmetric Gaussian
// PSFs only (h^T == h, so we reuse the same Ĥ for both forward and
// back-projection convolutions). epsilon = 1e-6 in the ratio guards
// division-by-zero in flat regions.

import { fft2, ifft2 } from './fft2';
import { padAndCenterPSF } from './image-psf';

export interface RichardsonLucyArgs {
  plane: Float32Array;   // length w*h
  w: number;             // power of 2
  h: number;             // power of 2
  psf: Float32Array;     // length psfW*psfH
  psfW: number;          // positive odd integer
  psfH: number;          // positive odd integer
  iterations?: number;   // default 20; must be >= 1
  epsilon?: number;      // default 1e-6
}

const DEFAULT_ITERATIONS = 20;
const DEFAULT_EPSILON = 1e-6;

function isPow2(n: number): boolean {
  return Number.isInteger(n) && n > 0 && (n & (n - 1)) === 0;
}

function isPositiveOdd(n: number): boolean {
  return Number.isInteger(n) && n > 0 && (n & 1) === 1;
}

/**
 * Handles the gnosis richardson Lucy Deblur workflow.
 */
export function richardsonLucyDeblur(args: RichardsonLucyArgs): Float32Array {
  const { plane, w, h, psf, psfW, psfH } = args;
  const iterations = args.iterations ?? DEFAULT_ITERATIONS;
  const epsilon = args.epsilon ?? DEFAULT_EPSILON;

  // Validate plane dimensions.
  if (!isPow2(w)) {
    throw new RangeError(`richardsonLucyDeblur: w must be a positive power of 2 (got ${w})`);
  }
  if (!isPow2(h)) {
    throw new RangeError(`richardsonLucyDeblur: h must be a positive power of 2 (got ${h})`);
  }
  if (plane.length !== w * h) {
    throw new RangeError(
      `richardsonLucyDeblur: plane.length (${plane.length}) must equal w*h (${w * h})`,
    );
  }

  // Validate PSF dimensions.
  if (!isPositiveOdd(psfW)) {
    throw new RangeError(
      `richardsonLucyDeblur: psfW must be a positive odd integer (got ${psfW})`,
    );
  }
  if (!isPositiveOdd(psfH)) {
    throw new RangeError(
      `richardsonLucyDeblur: psfH must be a positive odd integer (got ${psfH})`,
    );
  }
  if (psf.length !== psfW * psfH) {
    throw new RangeError(
      `richardsonLucyDeblur: psf.length (${psf.length}) must equal psfW*psfH (${psfW * psfH})`,
    );
  }
  if (psfW > w || psfH > h) {
    throw new RangeError(
      `richardsonLucyDeblur: PSF (${psfW}x${psfH}) must fit within plane (${w}x${h})`,
    );
  }

  // Validate iterations.
  if (!Number.isInteger(iterations) || iterations < 1) {
    throw new RangeError(
      `richardsonLucyDeblur: iterations must be an integer >= 1 (got ${iterations})`,
    );
  }

  // Validate epsilon.
  if (!Number.isFinite(epsilon) || epsilon <= 0) {
    throw new RangeError(
      `richardsonLucyDeblur: epsilon must be a positive finite number (got ${epsilon})`,
    );
  }

  // 1. Pad and center PSF to plane dimensions.
  const padded = padAndCenterPSF(psf, psfW, psfH, w, h);

  // 2. PSF spectrum (computed once; symmetric Gaussian => Ĥ used for both
  //    forward convolution and back-projection since h^T == h).
  const H = fft2(padded, w, h);

  const n = w * h;

  // 3. Initial guess x_0 = plane.
  const x = new Float32Array(plane);

  // 4. Iterate.
  for (let it = 0; it < iterations; it++) {
    // Forward conv: HX = IFFT(H · FFT(x)).
    const X = fft2(x, w, h);
    const HXre = new Float32Array(n);
    const HXim = new Float32Array(n);
    for (let i = 0; i < n; i++) {
      const ar = H.re[i]!;
      const ai = H.im[i]!;
      const br = X.re[i]!;
      const bi = X.im[i]!;
      HXre[i] = ar * br - ai * bi;
      HXim[i] = ar * bi + ai * br;
    }
    const HX = ifft2({ re: HXre, im: HXim, w, h });

    // Ratio: plane ./ (HX + epsilon).
    const ratio = new Float32Array(n);
    for (let i = 0; i < n; i++) {
      ratio[i] = plane[i]! / (HX[i]! + epsilon);
    }

    // Back-projection: corr = IFFT(H · FFT(ratio))  (h^T == h for symmetric Gaussian).
    const R = fft2(ratio, w, h);
    const Cre = new Float32Array(n);
    const Cim = new Float32Array(n);
    for (let i = 0; i < n; i++) {
      const ar = H.re[i]!;
      const ai = H.im[i]!;
      const br = R.re[i]!;
      const bi = R.im[i]!;
      Cre[i] = ar * br - ai * bi;
      Cim[i] = ar * bi + ai * br;
    }
    const corr = ifft2({ re: Cre, im: Cim, w, h });

    // Multiplicative update: x_{k+1} = x_k * corr.
    for (let i = 0; i < n; i++) {
      x[i] = x[i]! * corr[i]!;
    }
  }

  return x;
}
