// @a0n/gnosis/distributed-inference-host/src/image-bilateral.ts
// Edge-preserving bilateral filter (Tomasi & Manduchi 1998). Used as a
// post-pass on recovered Y planes to suppress Wiener ringing in flat
// regions without softening real edges. Spatial sigma controls reach;
// intensity sigma controls the "edge" threshold (deltas larger than ~3*sigma_i
// contribute almost nothing). Default sigma_spatial=2 px, sigma_intensity=0.05
// suit unit-range [0,1] luminance. Boundary handled by edge-replication.
// Cost is O(W*H*(2r+1)^2); ~46M ops for 512x512 with r=6 (~1s in JS).

export interface BilateralFilterArgs {
  plane: Float32Array;       // length width * height; values in [0, 1]
  width: number;
  height: number;
  sigmaSpatial?: number;     // default 2.0 (pixels)
  sigmaIntensity?: number;   // default 0.05 (in input value units, e.g. [0,1])
  kernelRadius?: number;     // default ceil(3 * sigmaSpatial)
}

/**
 * Handles the gnosis bilateral Filter workflow.
 */
export function bilateralFilter(args: BilateralFilterArgs): Float32Array {
  const { plane, width, height } = args;
  const sigmaSpatial = args.sigmaSpatial ?? 2.0;
  const sigmaIntensity = args.sigmaIntensity ?? 0.05;
  const kernelRadius = args.kernelRadius ?? Math.ceil(3 * sigmaSpatial);

  if (!(width > 0)) {
    throw new RangeError(`bilateralFilter: width must be > 0 (got ${width})`);
  }
  if (!(height > 0)) {
    throw new RangeError(`bilateralFilter: height must be > 0 (got ${height})`);
  }
  if (plane.length !== width * height) {
    throw new RangeError(
      `bilateralFilter: plane.length (${plane.length}) !== width*height (${width * height})`,
    );
  }
  if (!(sigmaSpatial > 0)) {
    throw new RangeError(
      `bilateralFilter: sigmaSpatial must be > 0 (got ${sigmaSpatial})`,
    );
  }
  if (!(sigmaIntensity > 0)) {
    throw new RangeError(
      `bilateralFilter: sigmaIntensity must be > 0 (got ${sigmaIntensity})`,
    );
  }
  if (!(kernelRadius >= 0) || !Number.isFinite(kernelRadius)) {
    throw new RangeError(
      `bilateralFilter: kernelRadius must be >= 0 (got ${kernelRadius})`,
    );
  }

  const r = kernelRadius | 0;
  const side = 2 * r + 1;

  // Pre-compute the spatial Gaussian weights as a (2r+1)*(2r+1) table indexed
  // by (dy + r) * side + (dx + r). The inner loops just multiply.
  const spatial = new Float32Array(side * side);
  const twoSigmaSpatialSq = 2 * sigmaSpatial * sigmaSpatial;
  for (let dy = -r; dy <= r; dy++) {
    for (let dx = -r; dx <= r; dx++) {
      const d2 = dx * dx + dy * dy;
      spatial[(dy + r) * side + (dx + r)] = Math.exp(-d2 / twoSigmaSpatialSq);
    }
  }

  const twoSigmaIntensitySq = 2 * sigmaIntensity * sigmaIntensity;
  const out = new Float32Array(width * height);

  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      const center = plane[y * width + x];
      let sum = 0;
      let wsum = 0;

      for (let dy = -r; dy <= r; dy++) {
        let sy = y + dy;
        // Replicate boundary
        if (sy < 0) sy = 0;
        else if (sy >= height) sy = height - 1;
        const rowBase = sy * width;
        const spatialRowBase = (dy + r) * side;

        for (let dx = -r; dx <= r; dx++) {
          let sx = x + dx;
          if (sx < 0) sx = 0;
          else if (sx >= width) sx = width - 1;

          const sample = plane[rowBase + sx];
          const ws = spatial[spatialRowBase + (dx + r)];
          const delta = sample - center;
          const wi = Math.exp(-(delta * delta) / twoSigmaIntensitySq);
          const w = ws * wi;
          sum += sample * w;
          wsum += w;
        }
      }

      out[y * width + x] = wsum > 0 ? sum / wsum : center;
    }
  }

  return out;
}
