// @a0n/gnosis/distributed-inference-host/src/impulsive-event-admission.ts
/**
 * # Impulsive Event Admission — sharp-onset detection with cross-node quorum
 *
 * Mirrors `Gnosis.Mesh.ImpulsiveEventAdmission` from `open-source/gnosis-math`,
 * the Rustic Church carrier for the impulsive acoustic event problem: a
 * glass-break or a gunshot is an **impulsive** event — a very sharp onset (a
 * large energy rise concentrated in one or two frames) followed by a decay. A
 * sustained sound (a held note, an HVAC hum, applause) rises *gradually*: its
 * energy climbs over many frames, so no single frame carries a large rise.
 *
 * The Skeptic admits an "impulsive event" claim only when two gates clear:
 *
 *  1. **Onset-sharpness gate**: the largest single-frame energy rise across the
 *     window must exceed a sharpness threshold. A sharp impulse has one big jump;
 *     a sustained sound, even one reaching the same total energy, only ever rises
 *     a little per frame, so it never clears the gate. This is the onset-shape
 *     discriminator and the module's hard negative.
 *  2. **Cross-node quorum gate**: a quorum `K >= 2` of independent mesh nodes
 *     must each report a sharp onset within the tight detection window. One
 *     microphone can be fooled (a knock on the casing, a clipped artifact, an
 *     injection), so a single-node report is held back — it can never certify on
 *     its own. This reuses the single-ear refutation of
 *     `./cross-node-coincidence`.
 *
 * So a sustained sound (gradual onset, low sharpness) is rejected outright, and a
 * genuinely sharp event heard by only one node is held back until the mesh
 * agrees. The hard negative is an impulsive and a sustained frame list carrying
 * the **same total energy** — only the onset *shape* distinguishes them; source
 * loudness alone cannot serve as an impulse key.
 *
 * ## What is (and is not) carried
 *
 * This module is the runtime promotion check for an impulsive-event observation.
 * It carries ONLY the finite contract the Lean file proves: the saturating onset
 * sharpness (`onsetSharpness`), the sharp-onset count over node reports
 * (`countSharp`), and quorum admission (`admitImpulse`). Real STFT onset
 * envelopes, true energy estimation, multipath and reverberation, node clock
 * synchronization, and classifier robustness stay runtime/empirical obligations
 * OUTSIDE this certificate.
 *
 * The promotion check folds the contract into the same epistemic-outcome
 * vocabulary the sibling mesh carriers already speak
 * (`claim_admitted` / `observation_only` / `claim_withheld`), so an
 * impulsive-event claim cannot be displayed until it has cleared the
 * Lean-side gates.
 *
 * ## Lean source-of-truth
 *
 * @see open-source/gnosis-math/Gnosis/Mesh/ImpulsiveEventAdmission.lean
 * @see open-source/gnosis-math/Gnosis/Mesh/CrossNodeCoincidence.lean (quorum sibling)
 */

/* ------------------------------------------------------------------ */
/*  Onset sharpness                                                   */
/* ------------------------------------------------------------------ */

/**
 * The largest single-frame energy rise across a window: the max over every
 * successive pair `(cur, next)` of the saturating rise `next - cur`. A sharp
 * impulse has one big jump; a sustained sound only ever rises a little per frame.
 * Saturating subtraction means a *decay* (`next < cur`) contributes `0` — only
 * rises count. An empty or single-frame window has zero sharpness.
 * Mirrors `ImpulsiveEventAdmission.onsetSharpness`.
 */
export function onsetSharpness(frames: readonly number[]): number {
  let maxRise = 0;
  for (let i = 0; i + 1 < frames.length; i++) {
    const rise = Math.max(0, frames[i + 1] - frames[i]);
    if (rise > maxRise) {
      maxRise = rise;
    }
  }
  return maxRise;
}

/* ------------------------------------------------------------------ */
/*  Node reports and the sharp-onset count                            */
/* ------------------------------------------------------------------ */

/**
 * A mesh node's report on a candidate impulsive event: the onset sharpness it
 * measured and whether it detected anything in the window.
 * Mirrors `ImpulsiveEventAdmission.NodeReport`.
 */
export interface NodeReport {
  readonly sharpness: number;
  readonly detected: boolean;
}

/**
 * How many detected reports clear the sharpness threshold. A gradual (sustained)
 * report does not clear the threshold and contributes nothing, so it cannot
 * inflate the quorum. Mirrors `ImpulsiveEventAdmission.countSharp`.
 */
export function countSharp(
  threshold: number,
  reports: readonly NodeReport[],
): number {
  let count = 0;
  for (const r of reports) {
    if (r.detected && r.sharpness >= threshold) {
      count += 1;
    }
  }
  return count;
}

/* ------------------------------------------------------------------ */
/*  Runtime promotion check (epistemic outcome)                       */
/* ------------------------------------------------------------------ */

/**
 * Epistemic outcome of an impulsive-event observation. Identical vocabulary to
 * the sibling mesh carriers, kept as a local literal union so this certificate
 * has no import coupling:
 *
 *  - `claim_admitted`    — a quorum `K >= 2` of independent nodes each report a
 *                          sharp onset, so the impulse may be promoted to a real
 *                          event claim.
 *  - `observation_only`  — one or more nodes report a sharp onset but the quorum
 *                          is not met (an unquorate, genuinely sharp detection),
 *                          OR the requested quorum is `K < 2` and so cannot
 *                          self-certify; reported, never promoted.
 *  - `claim_withheld`    — no node reports a sharp onset (a sustained sound, or
 *                          nothing impulsive at all), so there is no impulse to
 *                          claim.
 */
export type ImpulseEpistemicOutcome =
  | 'claim_admitted'
  | 'observation_only'
  | 'claim_withheld';

export interface ImpulseAdmission {
  readonly outcome: ImpulseEpistemicOutcome;
  /** How many nodes reported a sharp onset clearing the sharpness threshold. */
  readonly sharpCount: number;
  /** Whether the quorum was met (sharpCount >= K AND K >= 2). */
  readonly quorumMet: boolean;
  readonly summary: string;
}

export interface ImpulseAdmitOptions {
  /** The onset-sharpness a node report must clear to count as a sharp onset. */
  readonly sharpThreshold: number;
  /** The number of sharp-onset nodes required to promote (must be >= 2 to inherit
   *  the single-node refutation). */
  readonly quorumK: number;
}

/**
 * The promotion check Monster Mesh runs before any UI surface displays an
 * impulsive event (a glass-break, a gunshot) as a real event. It folds the Lean
 * contract into an epistemic outcome: the sharp-onset count decides whether
 * there is anything to claim, and the quorum `K >= 2` decides whether the claim
 * is allowed to be a *real-event* claim rather than the report of a lone node
 * that cannot self-certify.
 *
 * A quorum of `K < 2` is refused outright (mirrors how the Lean module forbids
 * quorum-1 admission via `single_node_no_impulse`): a single node — real, noisy,
 * or spoofed — cannot manufacture a second independent confirmation, so it never
 * reaches a promotable claim.
 *
 * @param reports the per-node onset reports for this candidate event
 * @param opts    the sharpness threshold and the required quorum K
 */
export function admitImpulse(
  reports: readonly NodeReport[],
  opts: ImpulseAdmitOptions,
): ImpulseAdmission {
  const sharpCount = countSharp(opts.sharpThreshold, reports);

  // A quorum below two cannot inherit the Lean single-node refutation
  // (`single_node_no_impulse` forbids admission at K = 1).
  const quorumIsSane = opts.quorumK >= 2;
  const quorumMet = quorumIsSane && sharpCount >= opts.quorumK;

  if (sharpCount === 0) {
    return {
      outcome: 'claim_withheld',
      sharpCount,
      quorumMet: false,
      summary:
        'no node reports a sharp onset — sustained or absent, nothing impulsive to claim',
    };
  }

  if (quorumMet) {
    return {
      outcome: 'claim_admitted',
      sharpCount,
      quorumMet: true,
      summary: `quorum K=${opts.quorumK} met: ${sharpCount} independent nodes each report a sharp onset`,
    };
  }

  return {
    outcome: 'observation_only',
    sharpCount,
    quorumMet: false,
    summary: quorumIsSane
      ? `unquorate impulse: ${sharpCount} sharp-onset node(s) below quorum K=${opts.quorumK}`
      : `quorum K=${opts.quorumK} < 2 cannot self-certify a single node`,
  };
}

/* ------------------------------------------------------------------ */
/*  Concrete fixtures (mirror the Lean fixtures)                      */
/* ------------------------------------------------------------------ */

/** The sharpness threshold used throughout the fixtures. Mirrors the Lean `40`. */
export const sharpThreshold = 40;

/** An impulsive window: a sharp 87-frame onset jump (`3 -> 90`) then decay.
 *  Successive rises 1, 87, 0 (decay), 0 (decay) → max rise 87. Total energy 145.
 *  Mirrors `ImpulsiveEventAdmission.impulseFrames`. */
export const impulseFrames: readonly number[] = [2, 3, 90, 40, 10];

/** A sustained window, near-identical total energy: a steady gradual climb.
 *  Successive rises all 10 → max rise 10. Total energy 150.
 *  Mirrors `ImpulsiveEventAdmission.sustainedFrames`. */
export const sustainedFrames: readonly number[] = [10, 20, 30, 40, 50];

/** Three independent nodes each reporting a sharp onset (87, 80, 75 ≥ 40).
 *  Mirrors `ImpulsiveEventAdmission.threeSharp`. */
export const threeSharp: readonly NodeReport[] = [
  { sharpness: 87, detected: true },
  { sharpness: 80, detected: true },
  { sharpness: 75, detected: true },
];

/** A single node reporting a sharp onset — genuine, but alone.
 *  Mirrors `ImpulsiveEventAdmission.oneSharp`. */
export const oneSharp: readonly NodeReport[] = [{ sharpness: 87, detected: true }];

/** A node reporting only a gradual (sustained) onset: not a sharp event.
 *  Mirrors `ImpulsiveEventAdmission.sustainedReport`. */
export const sustainedReport: NodeReport = { sharpness: 10, detected: true };

/** A sharp node and a sustained node together: only the sharp one counts.
 *  Mirrors `ImpulsiveEventAdmission.mixed`. */
export const mixed: readonly NodeReport[] = [
  { sharpness: 87, detected: true },
  { sharpness: 10, detected: true },
];

/**
 * Impulsive Event master witness. Bundles the boundary facts the Lean theorem
 * `impulsive_event_master` proves, each recomputed here from first principles
 * (never imported from Lean IR).
 *
 * Mirrors `ImpulsiveEventAdmission.impulsive_event_master`.
 */
export const ImpulsiveEventMaster = {
  /** The impulsive window's onset sharpness is exactly the 87-frame jump. */
  impulseSharpness: onsetSharpness(impulseFrames) === 87,

  /** The impulsive window clears the sharpness threshold (87 >= 40). */
  impulseFramesSharp: onsetSharpness(impulseFrames) >= sharpThreshold,

  /** The sustained window's onset sharpness is small (max rise 10). */
  sustainedSharpness: onsetSharpness(sustainedFrames) === 10,

  /** The sustained window does NOT clear the sharpness threshold (10 < 40). */
  sustainedFramesNotSharp: onsetSharpness(sustainedFrames) < sharpThreshold,

  /** All three nodes report sharp onsets: the sharp count is three. */
  threeSharpCount: countSharp(sharpThreshold, threeSharp) === 3,

  /** The mesh admits the impulse at quorum two over the three sharp reports. */
  impulseAdmittedAtTwo:
    admitImpulse(threeSharp, { sharpThreshold, quorumK: 2 }).outcome ===
    'claim_admitted',

  /** And the coincidence is full: all three nodes agree (quorum three). */
  impulseAdmittedAtThree:
    admitImpulse(threeSharp, { sharpThreshold, quorumK: 3 }).outcome ===
    'claim_admitted',

  /** A single sharp node is held back as observation_only at quorum two. */
  oneSharpObservationOnly:
    admitImpulse(oneSharp, { sharpThreshold, quorumK: 2 }).outcome ===
    'observation_only',

  /** A single sharp node is NOT admitted at quorum two. */
  oneSharpNotAdmitted:
    admitImpulse(oneSharp, { sharpThreshold, quorumK: 2 }).outcome !==
    'claim_admitted',

  /** A sustained report is not counted as sharp: dropping it from the mixed list
   *  leaves the sharp count unchanged at one. A gradual onset cannot inflate the
   *  quorum. */
  sustainedNotCounted:
    countSharp(sharpThreshold, mixed) === countSharp(sharpThreshold, oneSharp),

  /** The standing hard negative: the impulsive window clears the gate and the
   *  sustained window does not, despite near-identical total energy. */
  impulseVsSustained:
    onsetSharpness(impulseFrames) >= sharpThreshold &&
    onsetSharpness(sustainedFrames) < sharpThreshold,
} as const;
