// @a0n/gnosis/distributed-inference-host/src/keystroke-leak-detector.ts
/**
 * # Keystroke Acoustic Leak Detector — a privacy WARNING carrier
 *
 * Mirrors `Gnosis.Mesh.KeystrokeLeakDetector` from `open-source/gnosis-math`,
 * the Rustic Church carrier for the keyboard-acoustics privacy hazard framed
 * **strictly defensively** as a leak detector that *warns a user*, never as a
 * key-recovery tool. The physical fact it models: different keyboard keys make
 * acoustically distinguishable sounds (different plate positions, travel, and
 * resonance), so a nearby microphone can LEAK *which* key was pressed. This
 * module does NOT decode keystrokes and carries no machinery to do so; it only
 * FLAGS that the environment is acoustically revealing, so the user can be
 * warned that their typing is exposed.
 *
 * The polarity is the INVERSE of `./microphone-fingerprint`: a same-device
 * provenance match wants two signatures to be CLOSE; the keystroke leak fires
 * when two DISTINCT keys are FAR apart, because that separability is exactly
 * what an eavesdropper could exploit to tell keys apart. Here "admitted" means
 * "a privacy LEAK is present here — keys are acoustically distinguishable, warn
 * the user", NOT that any key was decoded.
 *
 * ## What is (and is not) carried
 *
 * This module is the runtime promotion check for a keystroke-leak privacy
 * warning. It carries ONLY the finite contract the Lean file proves:
 *
 *  - the saturating per-band acoustic metric (`adist`);
 *  - the signature distance as the summed pointwise `adist`, with a large
 *    sentinel on a length mismatch (`fpDistance` / `mismatchPenalty`);
 *  - the leak gate as separation beyond `sepThreshold` (`keyLeaks`);
 *  - the safe gate as indistinguishability within `intraTol`;
 *  - the leak-warning verdict folded into the epistemic-outcome vocabulary
 *    (`assessKeystrokeLeak`).
 *
 * Real acoustic feature extraction, per-band spectral estimation, room
 * reverberation, microphone placement, and the actual statistical separability
 * of a keyboard's key population stay runtime/empirical obligations OUTSIDE this
 * certificate — they are not theorems and are not asserted here.
 *
 * The promotion check folds the contract into the same epistemic-outcome
 * vocabulary the sibling mesh carriers already speak
 * (`claim_admitted` / `observation_only` / `claim_withheld`), so a
 * "your typing may be acoustically exposed here" warning is raised only when the
 * Lean-side separability gate fires. STRICTLY DEFENSIVE throughout: this carrier
 * neither performs nor enables keystroke recovery.
 *
 * ## Lean source-of-truth
 *
 * @see open-source/gnosis-math/Gnosis/Mesh/KeystrokeLeakDetector.lean
 * @see open-source/gnosis-math/Gnosis/Mesh/MicrophoneFingerprint.lean (inverse-polarity sibling)
 */

/* ------------------------------------------------------------------ */
/*  Per-band distance and key acoustic signatures                     */
/* ------------------------------------------------------------------ */

/**
 * A key acoustic signature is the per-band magnitude vector of one keypress over
 * a fixed analysis window. Magnitudes are non-negative integers (saturating
 * `Nat` arithmetic), matching the Lean `KeyAcoustic`.
 */
export type KeyAcoustic = readonly number[];

/** Absolute difference on `Nat` via saturating subtraction (`|a - b|`). The
 *  per-band acoustic residual metric, shared with the mesh corridor carriers.
 *  Mirrors `KeystrokeLeakDetector.adist`. */
export function adist(a: number, b: number): number {
  return Math.max(0, a - b) + Math.max(0, b - a);
}

/**
 * Sentinel returned on a length mismatch: large enough that no sane threshold
 * treats it as a near match. Differently-windowed signatures are reported as a
 * large separation (which, for a leak detector, reads as a separated pair).
 * Mirrors `KeystrokeLeakDetector.mismatchPenalty`.
 */
export const mismatchPenalty = 1000000;

/**
 * The signature distance: the sum of pointwise `adist` between aligned bands.
 * On a length mismatch it returns `mismatchPenalty` rather than attempting a
 * partial alignment. Mirrors `KeystrokeLeakDetector.fpDistance`.
 */
export function fpDistance(x: KeyAcoustic, y: KeyAcoustic): number {
  if (x.length !== y.length) {
    return mismatchPenalty;
  }
  let sum = 0;
  for (let i = 0; i < x.length; i++) {
    sum += adist(x[i], y[i]);
  }
  return sum;
}

/* ------------------------------------------------------------------ */
/*  Leak gate (inter-key separability — THE WARNING)                  */
/* ------------------------------------------------------------------ */

/**
 * The privacy-leak warning predicate. Two key signatures are *separable* — hence
 * the environment LEAKS which key was pressed — when their signature distance
 * exceeds `sepThreshold`. This is the inverse of a fingerprint match: separation,
 * not closeness, is the hazard, because separability is exactly what lets a
 * nearby microphone tell two keys apart. Returns `true` to FLAG a leak (warn the
 * user); it does NOT identify the key. Mirrors `KeystrokeLeakDetector.leaksB` /
 * `LeakPresent`.
 *
 * @param keyX one keypress signature
 * @param keyY another keypress signature
 * @param opts the separability threshold the distance must exceed to be a leak
 */
export function keyLeaks(
  keyX: KeyAcoustic,
  keyY: KeyAcoustic,
  opts: { readonly sepThreshold: number },
): boolean {
  return fpDistance(keyX, keyY) > opts.sepThreshold;
}

/* ------------------------------------------------------------------ */
/*  Runtime promotion check (epistemic outcome)                       */
/* ------------------------------------------------------------------ */

/**
 * Epistemic outcome of a keystroke-leak privacy assessment. Identical vocabulary
 * to the sibling mesh carriers (`./microphone-fingerprint`), kept as a local
 * literal union so this certificate has no import coupling. Note the polarity:
 * `claim_admitted` here means a privacy LEAK is admitted as present (warn the
 * user), NOT that a key was decoded.
 *
 *  - `claim_admitted`    — the two key signatures separate beyond `sepThreshold`,
 *                          so the keys are acoustically distinguishable: a LEAK is
 *                          present and the user should be warned.
 *  - `observation_only`  — the separation sits in the ambiguous band (above
 *                          `intraTol`, but not beyond `sepThreshold`): possibly
 *                          distinguishable, not certifiably so — reported, never
 *                          promoted to a warning.
 *  - `claim_withheld`    — the signatures are within `intraTol`, i.e.
 *                          indistinguishable: no leak, nothing to warn about.
 */
export type KeystrokeLeakEpistemicOutcome =
  | 'claim_admitted'
  | 'observation_only'
  | 'claim_withheld';

/** The reason a leak warning was held back (or the safe verdict reached). */
export type KeystrokeLeakLimitationClass =
  | 'indistinguishable_safe'
  | 'ambiguous_margin'
  | 'none';

export interface KeystrokeLeakAssessment {
  readonly outcome: KeystrokeLeakEpistemicOutcome;
  readonly limitation: KeystrokeLeakLimitationClass;
  /** Whether the separability gate fired (a leak is present). */
  readonly leaks: boolean;
  /** Whether the pair is within `intraTol` (acoustically indistinguishable). */
  readonly safe: boolean;
  /** The measured signature distance (the displayable separation). */
  readonly distance: number;
  readonly summary: string;
}

export interface KeystrokeLeakAssessOptions {
  /** Intra-key tightness: at or below this distance the pair is indistinguishable. */
  readonly intraTol: number;
  /** Separability: beyond this distance the keys leak (are distinguishable). */
  readonly sepThreshold: number;
}

/**
 * The promotion check Monster Mesh runs before raising a "your typing may be
 * acoustically exposed here" privacy notice. It folds the Lean contract into an
 * epistemic outcome, with leak-detector polarity:
 *
 *  - separation beyond `sepThreshold` is the LEAK signal — the keys are
 *    distinguishable, so the warning is admitted (mirrors `distinct_keys_leak`);
 *  - a pair within `intraTol` is indistinguishable — no leak, withheld as SAFE
 *    (mirrors `indistinct_keys_safe` / `same_key_not_leak`);
 *  - the band in between is an ambiguous margin: reported as an observation, not
 *    promoted to a warning.
 *
 * This neither identifies nor decodes any keystroke; it only flags whether the
 * environment is acoustically revealing.
 *
 * @param keyX one keypress signature
 * @param keyY another keypress signature
 * @param opts the intra-key tolerance and the separability threshold
 */
export function assessKeystrokeLeak(
  keyX: KeyAcoustic,
  keyY: KeyAcoustic,
  opts: KeystrokeLeakAssessOptions,
): KeystrokeLeakAssessment {
  const distance = fpDistance(keyX, keyY);
  const leaks = distance > opts.sepThreshold;
  const safe = distance <= opts.intraTol;

  // Leak gate: separability beyond the threshold means the keys are
  // distinguishable — a privacy leak is present, warn the user.
  if (leaks) {
    return {
      outcome: 'claim_admitted',
      limitation: 'none',
      leaks,
      safe,
      distance,
      summary: `keys acoustically distinguishable (distance ${distance} > sepThreshold ${opts.sepThreshold}) — privacy LEAK present, warn the user`,
    };
  }

  // Safe gate: within the intra-key spread the keys are indistinguishable, so an
  // eavesdropper cannot tell them apart — nothing to warn about.
  if (safe) {
    return {
      outcome: 'claim_withheld',
      limitation: 'indistinguishable_safe',
      leaks,
      safe,
      distance,
      summary: `keys indistinguishable (distance ${distance} <= intraTol ${opts.intraTol}) — no leak, safe`,
    };
  }

  // Ambiguous margin: above the intra-key spread but not beyond the leak
  // threshold — possibly distinguishable, not certifiably so.
  return {
    outcome: 'observation_only',
    limitation: 'ambiguous_margin',
    leaks,
    safe,
    distance,
    summary: `separation ${distance} sits between intraTol ${opts.intraTol} and sepThreshold ${opts.sepThreshold} — ambiguous, reported but not promoted to a warning`,
  };
}

/* ------------------------------------------------------------------ */
/*  Concrete fixtures (mirror the Lean fixtures)                      */
/* ------------------------------------------------------------------ */

/** Press 1 of key 'A': its measured per-band acoustic signature (length 4).
 *  Mirrors `KeystrokeLeakDetector.keyA1`. */
export const keyA1: KeyAcoustic = [40, 22, 15, 9];

/** Press 2 of the SAME key 'A': near-identical, off by at most one unit per band.
 *  `fpDistance(keyA1, keyA2) = 1+1+0+1 = 3`, tight within `intraTol = 4` and far
 *  below `sepThreshold = 12`, so the two presses are NOT flagged as a leak.
 *  Mirrors `keyA2`. */
export const keyA2: KeyAcoustic = [41, 21, 15, 10];

/** A clearly DIFFERENT key 'L': its signature is far from key 'A'.
 *  `fpDistance(keyA1, keyL) = 28+17+23+18 = 86`, well beyond `sepThreshold = 12`,
 *  so the pair is separable and the environment LEAKS which key was pressed.
 *  Mirrors `keyL`. */
export const keyL: KeyAcoustic = [12, 5, 38, 27];

/** A degenerate, acoustically near-identical key, part 1. Mirrors `keyM`. */
export const keyM: KeyAcoustic = [20, 18, 11, 7];

/** Its near-identical neighbour: `fpDistance(keyM, keyN) = 0+0+0+1 = 1`, within
 *  `intraTol = 4` and far below `sepThreshold = 12`. Indistinguishable → SAFE →
 *  not flagged. Mirrors `keyN`. */
export const keyN: KeyAcoustic = [20, 18, 11, 8];

/**
 * Keystroke Leak master witness. Bundles the boundary facts the Lean theorem
 * `leak_master` proves, each recomputed here from first principles (never
 * imported from Lean IR).
 *
 * Mirrors `KeystrokeLeakDetector.leak_master`.
 */
export const KeystrokeLeakMaster = {
  /** A signature is at zero distance from itself (self-consistency limit). */
  fpDistanceSelfZero: fpDistance(keyA1, keyA1) === 0,

  /** The two presses of key 'A' cluster tightly within tolerance 4 (distance 3). */
  sameKeyTight:
    fpDistance(keyA1, keyA2) <= 4,

  /** Key 'A' and key 'L' separate by more than sepThreshold 12 (distance 86). */
  keysWellSeparated: fpDistance(keyA1, keyL) > 12,

  /** The separated key pair triggers the leak warning. */
  distinctKeysLeak:
    assessKeystrokeLeak(keyA1, keyL, { intraTol: 4, sepThreshold: 12 })
      .outcome === 'claim_admitted',

  /** The Boolean detector also fires on the separated pair. */
  distinctKeysLeakB: keyLeaks(keyA1, keyL, { sepThreshold: 12 }) === true,

  /** The same-key pair is NOT a leak (no separation between two presses). */
  sameKeyNotLeak: keyLeaks(keyA1, keyA2, { sepThreshold: 12 }) === false,

  /** The near-identical keys are indistinguishable within tolerance 4 (distance 1). */
  indistinctKeysSafe:
    assessKeystrokeLeak(keyM, keyN, { intraTol: 4, sepThreshold: 12 })
      .outcome === 'claim_withheld',

  /** And the detector does NOT fire on the indistinguishable pair. */
  indistinctKeysNotFlagged: keyLeaks(keyM, keyN, { sepThreshold: 12 }) === false,
} as const;
