// @a0n/gnosis/distributed-inference-host/src/knot-model-storage.ts
/**
 * KnotModelStorage — ModelStorage backed by a .knot file.
 *
 * The .knot tensor index already uses the same key format that
 * LocalLayerNodeClient and the GCP coordinator expect
 * (e.g. `blk_0_attn_q_weight.bin`, `token_embd_weight.bin`).
 *
 * On construction the entire file is mmap'd into a Node.js Buffer.
 * getChunk() returns a zero-copy ArrayBuffer slice (no extra allocations
 * for repeated calls to the same tensor).
 */

import { readFileSync } from 'node:fs';
import { parseKnotHeaderFromBytes } from './knot-header';
import type { KnotHeader } from './knot-header';
import type { TensorInfo } from './types';
import type { ModelStorage } from '@a0n/aether/src/portable/interfaces';

export class KnotModelStorage implements ModelStorage {
  private readonly buf: Buffer;
  private readonly header: KnotHeader;
  private readonly tensors: Record<string, TensorInfo>;
  private readonly payloadOffset: number;

  constructor(knotPath: string) {
    this.buf = readFileSync(knotPath);
    this.header = parseKnotHeaderFromBytes(
      new Uint8Array(this.buf.buffer, this.buf.byteOffset, this.buf.byteLength)
    );
    this.tensors = this.header.metadata.tensors as Record<string, TensorInfo>;
    this.payloadOffset = this.header.payloadOffset;
  }

  async getChunk(key: string): Promise<ArrayBuffer | null> {
    const info = this.tensors[key];
    if (!info) return null;
    const start = this.payloadOffset + info.offset;
    const end = start + info.length;
    // readFileSync Buffer is backed by an ArrayBuffer (not SharedArrayBuffer).
    const backing = this.buf.buffer.slice(
      this.buf.byteOffset + start,
      this.buf.byteOffset + end
    ) as ArrayBuffer;
    return backing;
  }

  async listKeys(prefix?: string): Promise<string[]> {
    const keys = Object.keys(this.tensors);
    if (prefix === undefined || prefix === '') return keys;
    return keys.filter((k) => k.startsWith(prefix));
  }

  async exists(key: string): Promise<boolean> {
    return Object.hasOwn(this.tensors, key);
  }

  /** Tensor metadata for a key, when present. */
  tensorInfo(key: string): TensorInfo | undefined {
    return this.tensors[key];
  }

  async getText(key: string): Promise<string | null> {
    // Knot files don't store text blobs; return null for all text lookups.
    // The coordinator falls back gracefully when manifest.json is absent.
    return null;
  }

  async getJson<T>(key: string): Promise<T | null> {
    return null;
  }

  async putChunk(_key: string, _data: ArrayBuffer): Promise<void> {
    throw new Error('KnotModelStorage is read-only');
  }

  async putText(_key: string, _data: string): Promise<void> {
    throw new Error('KnotModelStorage is read-only');
  }

  async putJson(_key: string, _data: unknown): Promise<void> {
    throw new Error('KnotModelStorage is read-only');
  }

  async deleteChunk(_key: string): Promise<void> {
    throw new Error('KnotModelStorage is read-only');
  }

  /** Returns all tensor keys available in this knot. */
  tensorKeys(): string[] {
    return Object.keys(this.tensors);
  }

  /** Config block from knot metadata. */
  get knotConfig(): Record<string, unknown> {
    return (this.header.metadata.config ?? {}) as Record<string, unknown>;
  }
}
