// @a0n/gnosis/distributed-inference-host/src/light-model-artifact.ts
/**
 * Light model artifact support for Monster Mesh host surfaces.
 *
 * This is the browser-safe, dependency-light contract shared by
 * apps/brain-backwards-from-light and apps/monster-studio. It intentionally
 * models the retained optics/retina/cortical-wave state, not a claim that a
 * downloaded PNG can reconstruct the original source photograph.
 */

export const LIGHT_FIELD_WIDTH = 128;
export const LIGHT_FIELD_HEIGHT = 80;
export const LIGHT_MAX_WAVE_MODES = 32;
export const LIGHT_MODEL_BUNDLE_SCHEMA = 'ai.forkjoin.light.model-bundle';
export const LIGHT_MODEL_BUNDLE_VERSION = 1;
export const LIGHT_MODEL_DESCRIPTION =
  'optics-retina-cortical-wave-retained-state';

export interface LightFoveaPoint {
  readonly x: number;
  readonly y: number;
}

export interface LightTranslationSettings {
  readonly fovea: LightFoveaPoint;
  readonly exposure: number;
  readonly modeCount: number;
  readonly pupilMm: number;
  readonly focus: number;
  readonly glare: number;
}

export interface LightStandingWaveMode {
  readonly m: number;
  readonly n: number;
  readonly amplitude: number;
  readonly signedAmplitude: number;
  readonly omega: number;
  readonly phase: number;
  readonly color: readonly [number, number, number];
  readonly energy: number;
}

export interface LightTranslationMetrics {
  readonly meanLuminance: number;
  readonly opticalSpread: number;
  readonly rodLoad: number;
  readonly coneBalance: number;
  readonly fovealLoad: number;
  readonly opponentEnergy: number;
  readonly corticalEnergy: number;
  readonly dominantFrequency: number;
  readonly retainedEnergy: number;
}

export interface LightTranslationArtifactInput {
  readonly width: number;
  readonly height: number;
  readonly texture: Uint8Array;
  readonly modes: readonly LightStandingWaveMode[];
  readonly metrics: LightTranslationMetrics;
}

export interface LightModelBundle {
  readonly schema: typeof LIGHT_MODEL_BUNDLE_SCHEMA;
  readonly version: typeof LIGHT_MODEL_BUNDLE_VERSION;
  readonly sourceName: string;
  readonly createdAt: string;
  readonly model: {
    readonly description: typeof LIGHT_MODEL_DESCRIPTION;
    readonly width: number;
    readonly height: number;
    readonly settings: LightTranslationSettings;
    readonly textureRgbaBase64: string;
    readonly modes: readonly LightStandingWaveMode[];
    readonly metrics: LightTranslationMetrics;
  };
}

export interface LightReverseReconstruction {
  readonly width: number;
  readonly height: number;
  readonly sourceEstimateRgba: Uint8Array;
  readonly retinalRgba: Uint8Array;
  readonly corticalRgba: Uint8Array;
  readonly metrics: {
    readonly retainedEnergy: number;
    readonly modeLoss: number;
    readonly textureQuantization: number;
    readonly sourceAmbiguity: number;
  };
}

export interface LightMonsterMeshArtifact {
  readonly kind: 'light-model-bundle';
  readonly schema: typeof LIGHT_MODEL_BUNDLE_SCHEMA;
  readonly version: typeof LIGHT_MODEL_BUNDLE_VERSION;
  readonly sourceName: string;
  readonly createdAt: string;
  readonly width: number;
  readonly height: number;
  readonly modeCount: number;
  readonly dominantModes: readonly LightStandingWaveMode[];
  readonly metrics: {
    readonly retainedEnergy: number;
    readonly modeLoss: number;
    readonly textureQuantization: number;
    readonly sourceAmbiguity: number;
    readonly meanLuminance: number;
    readonly opticalSpread: number;
    readonly corticalEnergy: number;
  };
  readonly previews: {
    readonly sourceEstimateRgba: Uint8Array;
    readonly retinalRgba: Uint8Array;
    readonly corticalRgba: Uint8Array;
  };
  readonly bundle: LightModelBundle;
}

/**
 * Handles the gnosis clamp Light01 workflow.
 */
export function clampLight01(value: number): number {
  return Math.max(0, Math.min(1, value));
}

/**
 * Creates the Light Model Bundle.
 */
export function createLightModelBundle(
  translation: LightTranslationArtifactInput,
  settings: LightTranslationSettings,
  sourceName: string,
): LightModelBundle {
  return {
    schema: LIGHT_MODEL_BUNDLE_SCHEMA,
    version: LIGHT_MODEL_BUNDLE_VERSION,
    sourceName,
    createdAt: new Date().toISOString(),
    model: {
      description: LIGHT_MODEL_DESCRIPTION,
      width: translation.width,
      height: translation.height,
      settings,
      textureRgbaBase64: lightBytesToBase64(translation.texture),
      modes: translation.modes,
      metrics: translation.metrics,
    },
  };
}

/**
 * Parses the Light Model Bundle.
 */
export function parseLightModelBundle(json: string): LightModelBundle {
  const parsed: unknown = JSON.parse(json);
  if (!isRecord(parsed)) throw new Error('Model bundle must be a JSON object');
  if (parsed.schema !== LIGHT_MODEL_BUNDLE_SCHEMA) {
    throw new Error('Unsupported model bundle schema');
  }
  if (parsed.version !== LIGHT_MODEL_BUNDLE_VERSION) {
    throw new Error('Unsupported model bundle version');
  }
  if (typeof parsed.sourceName !== 'string') {
    throw new Error('Model bundle missing sourceName');
  }
  if (typeof parsed.createdAt !== 'string') {
    throw new Error('Model bundle missing createdAt');
  }
  if (!isRecord(parsed.model)) {
    throw new Error('Model bundle missing model payload');
  }
  const model = parsed.model;
  if (model.description !== LIGHT_MODEL_DESCRIPTION) {
    throw new Error('Model bundle has an unsupported model description');
  }
  if (model.width !== LIGHT_FIELD_WIDTH || model.height !== LIGHT_FIELD_HEIGHT) {
    throw new Error('Model bundle dimensions do not match this renderer');
  }
  if (typeof model.textureRgbaBase64 !== 'string') {
    throw new Error('Model bundle missing texture');
  }
  const texture = lightBase64ToBytes(model.textureRgbaBase64);
  if (texture.length !== LIGHT_FIELD_WIDTH * LIGHT_FIELD_HEIGHT * 4) {
    throw new Error('Model bundle texture has the wrong length');
  }
  const settings = parseLightSettings(model.settings);
  const modes = parseLightModes(model.modes);
  const metrics = parseLightMetrics(model.metrics);
  return {
    schema: LIGHT_MODEL_BUNDLE_SCHEMA,
    version: LIGHT_MODEL_BUNDLE_VERSION,
    sourceName: parsed.sourceName,
    createdAt: parsed.createdAt,
    model: {
      description: LIGHT_MODEL_DESCRIPTION,
      width: LIGHT_FIELD_WIDTH,
      height: LIGHT_FIELD_HEIGHT,
      settings,
      textureRgbaBase64: model.textureRgbaBase64,
      modes,
      metrics,
    },
  };
}

/**
 * Handles the gnosis reconstruct Light Model Bundle workflow.
 */
export function reconstructLightModelBundle(
  bundle: LightModelBundle,
): LightReverseReconstruction {
  const texture = lightBase64ToBytes(bundle.model.textureRgbaBase64);
  const sourceEstimateRgba = new Uint8Array(LIGHT_FIELD_WIDTH * LIGHT_FIELD_HEIGHT * 4);
  const retinalRgba = new Uint8Array(LIGHT_FIELD_WIDTH * LIGHT_FIELD_HEIGHT * 4);
  const corticalRgba = reconstructLightCorticalRgba(bundle.model.modes);
  let ambiguity = 0;

  for (let idx = 0; idx < LIGHT_FIELD_WIDTH * LIGHT_FIELD_HEIGHT; idx += 1) {
    const off = idx * 4;
    const opticalLuma = texture[off] / 255;
    const retinal = texture[off + 1] / 255;
    const opponent = texture[off + 2] / 255 - 0.5;
    const fovea = texture[off + 3] / 255;
    const chroma = opponent * (0.22 + fovea * 0.22);
    const red = clampLight01(opticalLuma + chroma * 1.25);
    const green = clampLight01(opticalLuma - Math.abs(chroma) * 0.58 + retinal * 0.08);
    const blue = clampLight01(opticalLuma - chroma * 0.92 + (1 - fovea) * 0.04);
    sourceEstimateRgba[off] = Math.round(red * 255);
    sourceEstimateRgba[off + 1] = Math.round(green * 255);
    sourceEstimateRgba[off + 2] = Math.round(blue * 255);
    sourceEstimateRgba[off + 3] = 255;

    retinalRgba[off] = Math.round(clampLight01(opticalLuma * 0.68 + fovea * 0.2) * 255);
    retinalRgba[off + 1] = Math.round(clampLight01(retinal * 0.85 + fovea * 0.18) * 255);
    retinalRgba[off + 2] = Math.round(clampLight01(0.32 + opponent * 0.44 + retinal * 0.22) * 255);
    retinalRgba[off + 3] = 255;
    ambiguity += Math.abs(opponent) * 0.36 + (1 - fovea) * 0.08;
  }

  const retainedEnergy = clampLight01(bundle.model.metrics.retainedEnergy);
  return {
    width: LIGHT_FIELD_WIDTH,
    height: LIGHT_FIELD_HEIGHT,
    sourceEstimateRgba,
    retinalRgba,
    corticalRgba,
    metrics: {
      retainedEnergy,
      modeLoss: clampLight01(1 - retainedEnergy),
      textureQuantization: 1 / 255,
      sourceAmbiguity: ambiguity / (LIGHT_FIELD_WIDTH * LIGHT_FIELD_HEIGHT),
    },
  };
}

/**
 * Handles the gnosis project Light Monster Mesh Artifact workflow.
 */
export function projectLightMonsterMeshArtifact(
  bundle: LightModelBundle,
): LightMonsterMeshArtifact {
  const reconstruction = reconstructLightModelBundle(bundle);
  return {
    kind: 'light-model-bundle',
    schema: LIGHT_MODEL_BUNDLE_SCHEMA,
    version: LIGHT_MODEL_BUNDLE_VERSION,
    sourceName: bundle.sourceName,
    createdAt: bundle.createdAt,
    width: bundle.model.width,
    height: bundle.model.height,
    modeCount: bundle.model.modes.length,
    dominantModes: bundle.model.modes.slice(0, 8),
    metrics: {
      retainedEnergy: reconstruction.metrics.retainedEnergy,
      modeLoss: reconstruction.metrics.modeLoss,
      textureQuantization: reconstruction.metrics.textureQuantization,
      sourceAmbiguity: reconstruction.metrics.sourceAmbiguity,
      meanLuminance: bundle.model.metrics.meanLuminance,
      opticalSpread: bundle.model.metrics.opticalSpread,
      corticalEnergy: bundle.model.metrics.corticalEnergy,
    },
    previews: {
      sourceEstimateRgba: reconstruction.sourceEstimateRgba,
      retinalRgba: reconstruction.retinalRgba,
      corticalRgba: reconstruction.corticalRgba,
    },
    bundle,
  };
}

/**
 * Parses the Light Monster Mesh Artifact.
 */
export function parseLightMonsterMeshArtifact(json: string): LightMonsterMeshArtifact {
  return projectLightMonsterMeshArtifact(parseLightModelBundle(json));
}

/**
 * Handles the gnosis light Bytes To Base64 workflow.
 */
export function lightBytesToBase64(bytes: Uint8Array): string {
  let binary = '';
  const chunkSize = 0x8000;
  for (let offset = 0; offset < bytes.length; offset += chunkSize) {
    const chunk = bytes.slice(offset, offset + chunkSize);
    binary += String.fromCharCode(...chunk);
  }
  return btoa(binary);
}

/**
 * Handles the gnosis light Base64 To Bytes workflow.
 */
export function lightBase64ToBytes(value: string): Uint8Array {
  const binary = atob(value);
  const bytes = new Uint8Array(binary.length);
  for (let idx = 0; idx < binary.length; idx += 1) {
    bytes[idx] = binary.charCodeAt(idx);
  }
  return bytes;
}

function reconstructLightCorticalRgba(
  modes: readonly LightStandingWaveMode[],
): Uint8Array {
  const rgba = new Uint8Array(LIGHT_FIELD_WIDTH * LIGHT_FIELD_HEIGHT * 4);
  const scalar = new Float32Array(LIGHT_FIELD_WIDTH * LIGHT_FIELD_HEIGHT);
  let maxAbs = 0.0001;
  for (let y = 0; y < LIGHT_FIELD_HEIGHT; y += 1) {
    for (let x = 0; x < LIGHT_FIELD_WIDTH; x += 1) {
      const u = (x + 0.5) / LIGHT_FIELD_WIDTH;
      const v = (y + 0.5) / LIGHT_FIELD_HEIGHT;
      let value = 0;
      for (const mode of modes) {
        const basis = Math.sin(mode.m * Math.PI * u) * Math.sin(mode.n * Math.PI * v);
        value += mode.signedAmplitude * basis;
      }
      const idx = y * LIGHT_FIELD_WIDTH + x;
      scalar[idx] = value;
      maxAbs = Math.max(maxAbs, Math.abs(value));
    }
  }
  for (let y = 0; y < LIGHT_FIELD_HEIGHT; y += 1) {
    for (let x = 0; x < LIGHT_FIELD_WIDTH; x += 1) {
      const idx = y * LIGHT_FIELD_WIDTH + x;
      const off = idx * 4;
      const u = (x + 0.5) / LIGHT_FIELD_WIDTH;
      const v = (y + 0.5) / LIGHT_FIELD_HEIGHT;
      let red = 0;
      let green = 0;
      let blue = 0;
      let weight = 0.0001;
      for (const mode of modes) {
        const basis = Math.sin(mode.m * Math.PI * u) * Math.sin(mode.n * Math.PI * v);
        const contribution = Math.abs(mode.signedAmplitude * basis);
        red += mode.color[0] * contribution;
        green += mode.color[1] * contribution;
        blue += mode.color[2] * contribution;
        weight += contribution;
      }
      const crest = Math.abs((scalar[idx] ?? 0) / maxAbs);
      rgba[off] = Math.round(clampLight01((red / weight) * (0.4 + crest * 0.75)) * 255);
      rgba[off + 1] = Math.round(clampLight01((green / weight) * (0.35 + crest * 0.72)) * 255);
      rgba[off + 2] = Math.round(clampLight01((blue / weight) * (0.48 + crest * 0.82)) * 255);
      rgba[off + 3] = 255;
    }
  }
  return rgba;
}

function parseLightSettings(value: unknown): LightTranslationSettings {
  if (!isRecord(value)) throw new Error('Model bundle settings must be an object');
  if (!isRecord(value.fovea)) throw new Error('Model bundle settings missing fovea');
  return {
    fovea: {
      x: finiteNumber(value.fovea.x, 'fovea.x'),
      y: finiteNumber(value.fovea.y, 'fovea.y'),
    },
    exposure: finiteNumber(value.exposure, 'exposure'),
    modeCount: Math.round(finiteNumber(value.modeCount, 'modeCount')),
    pupilMm: finiteNumber(value.pupilMm, 'pupilMm'),
    focus: finiteNumber(value.focus, 'focus'),
    glare: finiteNumber(value.glare, 'glare'),
  };
}

function parseLightModes(value: unknown): readonly LightStandingWaveMode[] {
  if (!Array.isArray(value)) throw new Error('Model bundle modes must be an array');
  return value.slice(0, LIGHT_MAX_WAVE_MODES).map((entry, index) => {
    if (!isRecord(entry)) throw new Error(`Mode ${index} must be an object`);
    if (!Array.isArray(entry.color) || entry.color.length !== 3) {
      throw new Error(`Mode ${index} color must have three channels`);
    }
    return {
      m: Math.round(finiteNumber(entry.m, `mode ${index} m`)),
      n: Math.round(finiteNumber(entry.n, `mode ${index} n`)),
      amplitude: finiteNumber(entry.amplitude, `mode ${index} amplitude`),
      signedAmplitude: finiteNumber(entry.signedAmplitude, `mode ${index} signedAmplitude`),
      omega: finiteNumber(entry.omega, `mode ${index} omega`),
      phase: finiteNumber(entry.phase, `mode ${index} phase`),
      color: [
        finiteNumber(entry.color[0], `mode ${index} color r`),
        finiteNumber(entry.color[1], `mode ${index} color g`),
        finiteNumber(entry.color[2], `mode ${index} color b`),
      ],
      energy: finiteNumber(entry.energy, `mode ${index} energy`),
    };
  });
}

function parseLightMetrics(value: unknown): LightTranslationMetrics {
  if (!isRecord(value)) throw new Error('Model bundle metrics must be an object');
  return {
    meanLuminance: finiteNumber(value.meanLuminance, 'meanLuminance'),
    opticalSpread: finiteNumber(value.opticalSpread, 'opticalSpread'),
    rodLoad: finiteNumber(value.rodLoad, 'rodLoad'),
    coneBalance: finiteNumber(value.coneBalance, 'coneBalance'),
    fovealLoad: finiteNumber(value.fovealLoad, 'fovealLoad'),
    opponentEnergy: finiteNumber(value.opponentEnergy, 'opponentEnergy'),
    corticalEnergy: finiteNumber(value.corticalEnergy, 'corticalEnergy'),
    dominantFrequency: finiteNumber(value.dominantFrequency, 'dominantFrequency'),
    retainedEnergy: finiteNumber(value.retainedEnergy, 'retainedEnergy'),
  };
}

function finiteNumber(value: unknown, label: string): number {
  if (typeof value !== 'number' || !Number.isFinite(value)) {
    throw new Error(`Model bundle has invalid ${label}`);
  }
  return value;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null;
}
