// @a0n/gnosis/distributed-inference-host/src/local-knot-runtime.ts
/**
 * LocalKnotRuntime — AgenticChatRuntime backed entirely by in-process
 * aether inference via LocalLayerNodeClient + KnotModelStorage.
 *
 * Eliminates all HTTP round-trips for auxiliary agentic purposes
 * (tool-review, intent-evaluation, tool-args, tool-retry).
 *
 * The .knot file is read once at construction. Inference runs fully
 * in-process using WASM SIMD kernels when available, with JS fallbacks.
 */

import LlamaTokenizer from 'llama-tokenizer-js';
import { LocalLayerNodeClient } from '@a0n/aether/src/local-layer-node-client';
import type { ModelConfig } from '@a0n/aether/src/portable/interfaces';
import { dequantizeQ4K, dequantizeQ6K } from '@a0n/aether/src/engine/q4-dequant';
import { KnotModelStorage } from './knot-model-storage';
import {
  renderMessagesForPrompt,
  type AgenticChatRuntime,
  type TextGenerationRequest,
  type TextGenerationResult,
} from './agentic-chat';

// Dequantize F16 → F32.
function f16ToF32(buf: ArrayBuffer): Float32Array {
  const src = new Uint16Array(buf);
  const out = new Float32Array(src.length);
  for (let i = 0; i < src.length; i++) {
    const h = src[i];
    const exp = (h >>> 10) & 0x1f;
    const mant = h & 0x3ff;
    const sign = h >>> 15 ? -1 : 1;
    if (exp === 0) {
      out[i] = sign * 5.960464477539063e-8 * mant;
    } else if (exp === 31) {
      out[i] = mant ? NaN : sign * Infinity;
    } else {
      out[i] = sign * Math.pow(2, exp - 15) * (1 + mant / 1024);
    }
  }
  return out;
}

// Dequantize a tensor buffer to F32, dispatching on the stored tensor type.
// Used for resident tensors (embedding, output norm, lm_head) that must be
// fully materialized as F32 for embedding lookup and lm_head argmax.
function bufToF32(buf: ArrayBuffer, tensorType: string, numElements: number): Float32Array {
  if (tensorType === 'F32') return new Float32Array(buf);
  if (tensorType === 'F16') return f16ToF32(buf);
  if (tensorType === 'Q4_K') return dequantizeQ4K(new Uint8Array(buf), numElements);
  if (tensorType === 'Q6_K') return dequantizeQ6K(new Uint8Array(buf), numElements);
  // Unknown quantization: fall through to raw bytes (will produce garbage but won't crash).
  return new Float32Array(buf);
}

export interface LocalKnotRuntimeOptions {
  knotPath: string;
  /** Max new tokens per auxiliary infer call (default: 64). */
  maxAuxTokens?: number;
  /** Stop token IDs (default: [1, 2]). */
  stopTokens?: number[];
  /** Model name reported in TextGenerationResult (default: local-knot). */
  modelName?: string;
}

export class LocalKnotRuntime implements AgenticChatRuntime {
  private readonly storage: KnotModelStorage;
  private readonly layerClient: LocalLayerNodeClient;
  private readonly modelConfig: ModelConfig;
  private readonly numLayers: number;
  private readonly maxAuxTokens: number;
  private readonly stopSet: Set<number>;
  private readonly modelName: string;

  // Resident tensors loaded once.
  private embeddingWeight: Float32Array | null = null;
  private outputNormWeight: Float32Array | null = null;
  private lmHeadWeight: Float32Array | null = null;
  private residentsLoaded = false;

  logger = console;

  constructor(opts: LocalKnotRuntimeOptions) {
    this.storage = new KnotModelStorage(opts.knotPath);
    this.maxAuxTokens = opts.maxAuxTokens ?? 64;
    this.stopSet = new Set(opts.stopTokens ?? [1, 2]);
    this.modelName = opts.modelName ?? 'local-knot';

    const cfg = this.storage.knotConfig;
    this.numLayers = (cfg['num_layers'] as number) ?? 22;

    this.modelConfig = {
      name:
        (cfg['model_name'] as string) ??
        (cfg['name'] as string) ??
        this.modelName,
      hiddenDim: (cfg['hidden_dim'] as number) ?? 2048,
      numLayers: this.numLayers,
      numHeads: (cfg['num_heads'] as number) ?? 32,
      numKvHeads: (cfg['num_kv_heads'] as number) ?? 4,
      headDim: (cfg['head_dim'] as number) ?? 64,
      intermediateSize: (cfg['intermediate_size'] as number) ?? 5632,
      vocabSize: (cfg['vocab_size'] as number) ?? 32000,
      maxSeqLength:
        (cfg['max_seq_length'] as number) ??
        (cfg['context_length'] as number) ??
        8192,
      ropeTheta: (cfg['rope_theta'] as number) ?? 10000,
      rmsNormEps: (cfg['rms_norm_eps'] as number) ?? 1e-5,
      architecture: (cfg['arch'] as string) ?? 'llama',
    };

    this.layerClient = new LocalLayerNodeClient(
      this.storage,
      this.modelConfig,
      0,
      this.numLayers - 1
    );
  }

  private async loadResidents(): Promise<void> {
    if (this.residentsLoaded) return;

    const { hiddenDim, vocabSize } = this.modelConfig;

    const load = async (key: string, numElements: number): Promise<Float32Array | null> => {
      const buf = await this.storage.getChunk(key);
      if (!buf) return null;
      const info = this.storage.tensorInfo(key);
      return bufToF32(buf, info?.tensor_type ?? 'F32', numElements);
    };

    this.embeddingWeight = await load('token_embd_weight.bin', vocabSize * hiddenDim);
    this.outputNormWeight = await load('output_norm_weight.bin', hiddenDim);
    this.lmHeadWeight =
      (await load('output_weight.bin', vocabSize * hiddenDim)) ?? this.embeddingWeight; // tied weights

    this.residentsLoaded = true;
  }

  private embed(tokenId: number): Float32Array {
    if (!this.embeddingWeight) throw new Error('embedding not loaded');
    const dim = this.modelConfig.hiddenDim;
    return this.embeddingWeight.slice(tokenId * dim, (tokenId + 1) * dim);
  }

  private rmsNorm(
    x: Float32Array,
    weight: Float32Array,
    eps: number
  ): Float32Array {
    const n = x.length;
    let sq = 0;
    for (let i = 0; i < n; i++) sq += x[i] * x[i];
    const rms = Math.sqrt(sq / n + eps);
    const out = new Float32Array(n);
    for (let i = 0; i < n; i++) out[i] = (x[i] / rms) * weight[i];
    return out;
  }

  private lmHead(hidden: Float32Array): number {
    if (!this.lmHeadWeight || !this.outputNormWeight)
      throw new Error('lmHead not loaded');
    const eps = this.modelConfig.rmsNormEps ?? 1e-5;
    const normed = this.rmsNorm(hidden, this.outputNormWeight, eps);
    const vocab = this.modelConfig.vocabSize;
    const dim = this.modelConfig.hiddenDim;
    // Greedy argmax over logits.
    let bestId = 0;
    let bestScore = -Infinity;
    for (let v = 0; v < vocab; v++) {
      let dot = 0;
      const base = v * dim;
      for (let d = 0; d < dim; d++)
        dot += this.lmHeadWeight[base + d] * normed[d];
      if (dot > bestScore) {
        bestScore = dot;
        bestId = v;
      }
    }
    return bestId;
  }

  async infer(request: TextGenerationRequest): Promise<TextGenerationResult> {
    await this.loadResidents();

    const rendered = renderMessagesForPrompt(request.messages);
    const prompt = `<s>[INST] ${rendered} [/INST]`;
    const promptTokens = LlamaTokenizer.encode(prompt, true, false) as number[];
    const maxNew = Math.min(
      request.maxTokens ?? this.maxAuxTokens,
      this.maxAuxTokens
    );

    const generated: number[] = [];
    const requestId = `local-${Date.now()}`;

    // Prefill: run prompt tokens through the model to build KV cache.
    let hidden = this.embed(promptTokens[0]);
    for (let pos = 0; pos < promptTokens.length; pos++) {
      if (pos > 0) hidden = this.embed(promptTokens[pos]);
      const result = await this.layerClient.forward(hidden, pos, requestId, {
        sequenceLength: pos + 1,
      });
      hidden =
        result.hiddenStates ??
        (result as unknown as { hidden_states: Float32Array }).hidden_states ??
        hidden;
    }

    // Decode: autoregressive generation.
    let pos = promptTokens.length;
    for (let step = 0; step < maxNew; step++) {
      const nextToken = this.lmHead(hidden);
      if (this.stopSet.has(nextToken)) break;
      generated.push(nextToken);
      hidden = this.embed(nextToken);
      const result = await this.layerClient.forward(hidden, pos, requestId, {
        sequenceLength: 1,
      });
      hidden =
        result.hiddenStates ??
        (result as unknown as { hidden_states: Float32Array }).hidden_states ??
        hidden;
      pos++;
    }

    const content = (LlamaTokenizer.decode(generated) as string).trim();
    return {
      content,
      promptTokens: promptTokens.length,
      completionTokens: generated.length,
      model: this.modelName,
    };
  }
}
