// @a0n/gnosis/distributed-inference-host/src/mel-spectrogram.ts
/**
 * Whisper-spec log-mel spectrogram extractor in pure TypeScript.
 *
 * Used by Pneuma (gnosis-pneuma) as the host-side audio preprocessor. The
 * mesh expects a (n_mels, n_frames) Float32Array as input to the encoder.
 * No native deps so this runs identically in CF Workers, Bun, Node, browser.
 *
 * Parameters match `transformers.WhisperFeatureExtractor` exactly:
 *   sample_rate = 16000, n_fft = 400, hop_length = 160, n_mels = 80,
 *   f_min = 0, f_max = 8000, log10 then clamp(max-8) then (+4)/4.
 *
 * Validation: cross-check against HF Python on a fixed wav; output should
 * match within 1e-3 elementwise.
 */

const SAMPLE_RATE = 16000;
const N_FFT = 400;
const HOP_LENGTH = 160;
const N_MELS = 80;
const N_SAMPLES = SAMPLE_RATE * 30; // 30-second pad target
const N_FRAMES = N_SAMPLES / HOP_LENGTH; // 3000

/** Hz → mel using the slaney scale (matches HF Whisper). */
function hzToMel(hz: number): number {
  // Slaney mel scale used by Whisper:
  // below 1000 Hz: linear; above: log
  const f_min = 0.0;
  const f_sp = 200.0 / 3.0;
  const min_log_hz = 1000.0;
  const min_log_mel = (min_log_hz - f_min) / f_sp;
  const logstep = Math.log(6.4) / 27.0;
  if (hz >= min_log_hz) return min_log_mel + Math.log(hz / min_log_hz) / logstep;
  return (hz - f_min) / f_sp;
}

function melToHz(mel: number): number {
  const f_min = 0.0;
  const f_sp = 200.0 / 3.0;
  const min_log_hz = 1000.0;
  const min_log_mel = (min_log_hz - f_min) / f_sp;
  const logstep = Math.log(6.4) / 27.0;
  if (mel >= min_log_mel) return min_log_hz * Math.exp(logstep * (mel - min_log_mel));
  return f_min + f_sp * mel;
}

/**
 * Build the mel filterbank: an (n_mels, n_fft/2 + 1) matrix of triangular
 * filters. Multiplied against the power spectrum to produce mel-energy.
 */
function buildMelFilterbank(): Float32Array {
  const n_freqs = N_FFT / 2 + 1; // 201
  const fb = new Float32Array(N_MELS * n_freqs);

  const f_min = 0.0;
  const f_max = SAMPLE_RATE / 2.0; // 8000

  const mel_min = hzToMel(f_min);
  const mel_max = hzToMel(f_max);
  const mel_points: number[] = [];
  for (let i = 0; i < N_MELS + 2; i++) {
    mel_points.push(mel_min + (i * (mel_max - mel_min)) / (N_MELS + 1));
  }
  const hz_points = mel_points.map(melToHz);

  // FFT bin frequencies
  const fft_freqs = new Float32Array(n_freqs);
  for (let i = 0; i < n_freqs; i++) {
    fft_freqs[i] = (i * SAMPLE_RATE) / N_FFT;
  }

  // Slaney normalization: each filter has unit area
  for (let m = 0; m < N_MELS; m++) {
    const lower = hz_points[m]!;
    const center = hz_points[m + 1]!;
    const upper = hz_points[m + 2]!;
    const enorm = 2.0 / (upper - lower);
    for (let k = 0; k < n_freqs; k++) {
      const f = fft_freqs[k]!;
      let w = 0.0;
      if (f >= lower && f <= center) w = (f - lower) / (center - lower);
      else if (f > center && f <= upper) w = (upper - f) / (upper - center);
      fb[m * n_freqs + k] = w * enorm;
    }
  }
  return fb;
}

let MEL_FB: Float32Array | null = null;
function getMelFilterbank(): Float32Array {
  if (!MEL_FB) MEL_FB = buildMelFilterbank();
  return MEL_FB;
}

/**
 * Radix-2 Cooley-Tukey FFT, in-place on real input.
 * Caller passes real signal of length n (power of 2); we return interleaved
 * complex result [re_0, im_0, re_1, im_1, ...] of length 2n.
 */
function fft(real: Float32Array): Float32Array {
  const n = real.length;
  if ((n & (n - 1)) !== 0) return dft(real);
  const out = new Float32Array(n * 2);
  for (let i = 0; i < n; i++) out[i * 2] = real[i]!;

  // Bit-reversal permutation
  let j = 0;
  for (let i = 0; i < n - 1; i++) {
    if (i < j) {
      const tr = out[i * 2]!;
      const ti = out[i * 2 + 1]!;
      out[i * 2] = out[j * 2]!;
      out[i * 2 + 1] = out[j * 2 + 1]!;
      out[j * 2] = tr;
      out[j * 2 + 1] = ti;
    }
    let m = n / 2;
    while (m >= 2 && j >= m) { j -= m; m /= 2; }
    j += m;
  }

  // Butterfly
  for (let stage = 1; stage <= Math.log2(n); stage++) {
    const m = 1 << stage;
    const wpr = Math.cos(-2 * Math.PI / m);
    const wpi = Math.sin(-2 * Math.PI / m);
    for (let k = 0; k < n; k += m) {
      let wr = 1.0, wi = 0.0;
      for (let s = 0; s < m / 2; s++) {
        const a = (k + s) * 2;
        const b = (k + s + m / 2) * 2;
        const outA = out[a]!;
        const outAi = out[a + 1]!;
        const outB = out[b]!;
        const outBi = out[b + 1]!;
        const tr = wr * outB - wi * outBi;
        const ti = wr * outBi + wi * outB;
        out[b] = outA - tr;
        out[b + 1] = outAi - ti;
        out[a] = outA + tr;
        out[a + 1] = outAi + ti;
        const wr_next = wr * wpr - wi * wpi;
        wi = wr * wpi + wi * wpr;
        wr = wr_next;
      }
    }
  }
  return out;
}

let DFT_BASIS: Float32Array | null = null;
function getDftBasis(n: number): Float32Array {
  const nFreqs = Math.floor(n / 2) + 1;
  const expected = nFreqs * n * 2;
  if (DFT_BASIS && DFT_BASIS.length === expected) return DFT_BASIS;
  const basis = new Float32Array(expected);
  for (let k = 0; k < nFreqs; k++) {
    for (let t = 0; t < n; t++) {
      const angle = (-2 * Math.PI * k * t) / n;
      const offset = (k * n + t) * 2;
      basis[offset] = Math.cos(angle);
      basis[offset + 1] = Math.sin(angle);
    }
  }
  DFT_BASIS = basis;
  return basis;
}

function dft(real: Float32Array): Float32Array {
  const n = real.length;
  const nFreqs = Math.floor(n / 2) + 1;
  const basis = getDftBasis(n);
  const out = new Float32Array(n * 2);
  for (let k = 0; k < nFreqs; k++) {
    let re = 0;
    let im = 0;
    const row = k * n * 2;
    for (let t = 0; t < n; t++) {
      const x = real[t]!;
      re += x * basis[row + t * 2]!;
      im += x * basis[row + t * 2 + 1]!;
    }
    out[k * 2] = re;
    out[k * 2 + 1] = im;
  }
  return out;
}

let HANN: Float32Array | null = null;
function hannWindow(n: number): Float32Array {
  if (HANN && HANN.length === n) return HANN;
  const w = new Float32Array(n);
  for (let i = 0; i < n; i++) w[i] = 0.5 * (1 - Math.cos((2 * Math.PI * i) / n));
  HANN = w;
  return w;
}

/**
 * Extract Whisper log-mel spectrogram from a 16 kHz mono float32 audio array.
 *
 * @param audio Mono float32 PCM at 16 kHz. Resample externally before calling.
 * @returns Float32Array of length n_mels * n_frames (80 * 3000), row-major
 *          with mel index outer, time index inner. Padded/truncated to 30 s.
 */
export function extractMelSpectrogram(audio: Float32Array): Float32Array {
  // Pad / truncate to 30 s
  const padded = new Float32Array(N_SAMPLES);
  const copyLen = Math.min(audio.length, N_SAMPLES);
  padded.set(audio.subarray(0, copyLen));
  const centered = reflectPad(padded, N_FFT / 2);

  const win = hannWindow(N_FFT);
  const n_freqs = N_FFT / 2 + 1;
  const stftFrames = N_FRAMES + 1;
  const power = new Float32Array(stftFrames * n_freqs);

  // STFT: hop through audio, window, FFT, power spectrum
  const frame = new Float32Array(N_FFT);
  for (let f = 0; f < stftFrames; f++) {
    const start = f * HOP_LENGTH;
    for (let i = 0; i < N_FFT; i++) {
      frame[i] = centered[start + i]! * win[i]!;
    }
    const spec = fft(frame);
    for (let k = 0; k < n_freqs; k++) {
      const re = spec[k * 2]!;
      const im = spec[k * 2 + 1]!;
      power[f * n_freqs + k] = re * re + im * im;
    }
  }

  // Mel projection: (n_frames, n_freqs) @ (n_mels, n_freqs)^T
  const fb = getMelFilterbank();
  const mel = new Float32Array(N_MELS * N_FRAMES);
  for (let m = 0; m < N_MELS; m++) {
    for (let f = 0; f < N_FRAMES; f++) {
      let s = 0;
      for (let k = 0; k < n_freqs; k++) s += power[f * n_freqs + k]! * fb[m * n_freqs + k]!;
      mel[m * N_FRAMES + f] = s;
    }
  }

  // log10, clamp at max-8, normalize to ~[-1, 1] (matches HF Whisper exactly)
  let maxVal = -Infinity;
  for (let i = 0; i < mel.length; i++) {
    const v = Math.log10(Math.max(mel[i]!, 1e-10));
    mel[i] = v;
    if (v > maxVal) maxVal = v;
  }
  const floor = maxVal - 8.0;
  for (let i = 0; i < mel.length; i++) {
    mel[i] = (Math.max(mel[i]!, floor) + 4.0) / 4.0;
  }
  return mel;
}

function reflectPad(input: Float32Array, pad: number): Float32Array {
  const out = new Float32Array(input.length + 2 * pad);
  for (let i = 0; i < pad; i++) {
    out[i] = input[pad - i] ?? 0;
  }
  out.set(input, pad);
  const tailStart = pad + input.length;
  for (let i = 0; i < pad; i++) {
    out[tailStart + i] = input[input.length - 2 - i] ?? 0;
  }
  return out;
}

/** Public constants, also for tests. */
export const WHISPER_MEL = {
  sampleRate: SAMPLE_RATE,
  nFft: N_FFT,
  hopLength: HOP_LENGTH,
  nMels: N_MELS,
  nFrames: N_FRAMES,
  nSamples: N_SAMPLES,
} as const;
