// @a0n/gnosis/distributed-inference-host/src/memetic-cipher-solver.ts
// File header (12-15 lines, MANDATORY): memetic algorithm — GA + local
// search via SA — for homophonic substitution ciphers. Each generation:
// (1) every individual gets SA-polished to its local optimum;
// (2) fitness-rank by combinedEnglishScore (mono+bi+quad blend) by
//     default — smoother gradient than pure quadgram, helps GA escape
//     plateaus. Set useCombinedScore:false to revert to pure quadgram.
//     NOTE: the inner SA polish (homophonic-substitution-solver) still
//     uses pure quadgram — that's the local search where smoothness is
//     less critical; the outer GA selection is where tie-breaking from
//     the combined score matters most;
// (3) Buleyean-weighted parent selection (rank-0 parent gets weight R+1);
// (4) two-point crossover; cut points driven by a logistic-map chaos
//     iterator, NOT plain LCG noise — the chaotic dynamics produce
//     non-uniform exploration that's empirically better than uniform
//     random for hard combinatorial spaces;
// (5) mutation rate also chaos-modulated; rises when fitness diversity
//     collapses (population becomes inbred), falls when diversity is high.
// The chaos-engine framing comes from open-source/gnosis/gnosis-chaos
// (Rust crate, FFI-only; we use a pure-JS logistic-map iterator that
// captures the same chaotic-evolution shape without the Rust binding).
//
// EXPORTS:
//   solveMemeticHomophonic        — synchronous, single-threaded path.
//                                   Stable API; existing tests/callers
//                                   stay on this entry point.
//   solveMemeticHomophonicParallel — async, runs the per-individual SA
//                                   polish step across worker_threads.
//                                   Returns Promise<MemeticSolverResult>.
//                                   Opt-in. Requires tsx as a workspace
//                                   dev dep (already present) so the
//                                   worker can load TS source directly
//                                   via execArgv: ['--import', 'tsx'].
//                                   Wall-time speedup ≈ min(workers,
//                                   populationSize) on a multi-core box.
//                                   No unit test in this task — the
//                                   sequential test covers correctness
//                                   and the parallel path is a pure
//                                   scheduling wrapper around it.

import {
  solveHomophonicSubstitution,
  type HomophonicSolverResult,
} from './homophonic-substitution-solver';
import { quadgramScore, combinedEnglishScore } from './english-frequency';
import { buleyeanWeight } from '@a0n/buleyean-kernel';
import { Worker } from 'node:worker_threads';
import os from 'node:os';
import { fileURLToPath } from 'node:url';
import { resolve as resolvePath, dirname } from 'node:path';

export interface MemeticSolverArgs {
  ciphertext: string;
  /** Population size. Default 24. */
  populationSize?: number;
  /** Number of generations. Default 30. */
  generations?: number;
  /** SA budget per individual per generation. Default 5000 swaps. */
  saSwapsPerIndividual?: number;
  /** SA temperature start / end / cooling. Forwarded to inner SA. */
  saTemperatureStart?: number;
  saTemperatureEnd?: number;
  saCoolingRate?: number;
  /** Mutation rate base (logistic map output is multiplied by this).
   *  Default 0.10 = up to 10% of symbols mutated per child. */
  mutationRateBase?: number;
  /** Logistic-map parameter. r = 3.9 is fully chaotic (r in (3.5699, 4]).
   *  Default 3.99. */
  logisticR?: number;
  /** Random seed. Default 0xC0DE_5EED. */
  seed?: number;
  /** Use the combined n-gram score (mono + bi + quad blend) for fitness
   *  evaluation. Default true — production-tuning empirically found this
   *  smooths the search landscape and helps escape pure-quadgram plateaus.
   *  Set to false to use pure quadgramScore (the legacy default). */
  useCombinedScore?: boolean;
  /** Per-generation progress callback. */
  onGeneration?: (info: GenerationInfo) => void;
  /** Run the per-individual SA polish step across worker_threads.
   *  Default: undefined (single-threaded). When set, spawns a pool of
   *  `workers` worker_threads (default = max(1, floor(os.cpus().length / 2)))
   *  and dispatches SA polish tasks to them per generation.
   *
   *  Honest scope: requires tsx as a dev dependency at the workspace root
   *  (already present in this repo). The Worker constructor uses
   *  `execArgv: ['--import', 'tsx']` so the worker process loads TS source
   *  directly. If your environment doesn't support this, leave `parallel`
   *  unset to use the single-threaded path.
   *
   *  Only honoured by `solveMemeticHomophonicParallel`. The synchronous
   *  `solveMemeticHomophonic` ignores this field. */
  parallel?: {
    workers?: number;
  };
}

export interface GenerationInfo {
  generation: number;
  bestScore: number;
  worstScore: number;
  meanScore: number;
  diversity: number;       // population score variance — drops over time
  mutationRate: number;    // chaos-driven, varies per generation
  bestPlaintext: string;
}

export interface MemeticSolverResult {
  plaintext: string;
  mapping: ReadonlyMap<string, string>;
  /** Fitness of the best-ever individual. Reports whatever score function
   *  was selected by `useCombinedScore`: combinedEnglishScore (default,
   *  true) or quadgramScore (legacy, false). Absolute values differ
   *  between the two — combined score is more compressed in range. */
  score: number;
  generations: number;
  finalPopulationStats: { bestScore: number; worstScore: number; meanScore: number };
  /** History of bestScore per generation. */
  scoreHistory: ReadonlyArray<number>;
  symbols: ReadonlyArray<string>;
}

// --- Logistic-map chaos iterator -----------------------------------------
// Captures the same chaotic-evolution shape as gnosis-chaos (Rust) without
// the FFI binding — pure JS, deterministic given (initial, r).
function logisticMap(initial: number, r: number): () => number {
  let x = initial;
  return () => {
    x = r * x * (1 - x);
    // Clamp to (0, 1) — numerical stability for r close to 4.
    if (x <= 0) x = 0.0001;
    if (x >= 1) x = 0.9999;
    return x;
  };
}

// --- Tiny seeded LCG for deterministic helpers ---------------------------
function makeLcg(seed: number): () => number {
  let s = (seed >>> 0) || 1;
  return () => {
    s = (Math.imul(s, 1664525) + 1013904223) >>> 0;
    return s / 0x1_0000_0000;
  };
}

// --- Population helpers --------------------------------------------------
interface Individual {
  mapping: Map<string, string>;
  score: number;
}

function cloneMapping(m: ReadonlyMap<string, string>): Map<string, string> {
  return new Map(m);
}

function applyMapping(
  ciphertext: string,
  mapping: ReadonlyMap<string, string>,
): string {
  let out = '';
  for (let i = 0; i < ciphertext.length; i++) {
    const ch = ciphertext[i]!;
    const mapped = mapping.get(ch);
    out += mapped ?? ch;
  }
  return out;
}

function scoreMapping(
  ciphertext: string,
  mapping: ReadonlyMap<string, string>,
  scoreFn: (s: string) => number,
): number {
  return scoreFn(applyMapping(ciphertext, mapping));
}

function meanOf(xs: ReadonlyArray<number>): number {
  if (xs.length === 0) return 0;
  let s = 0;
  for (const x of xs) s += x;
  return s / xs.length;
}

function stddevOf(xs: ReadonlyArray<number>): number {
  if (xs.length === 0) return 0;
  const m = meanOf(xs);
  let acc = 0;
  for (const x of xs) {
    const d = x - m;
    acc += d * d;
  }
  return Math.sqrt(acc / xs.length);
}

// --- Roulette over Buleyean weights --------------------------------------
function pickIndexByWeight(
  weights: ReadonlyArray<number>,
  rand: () => number,
): number {
  let total = 0;
  for (const w of weights) total += w;
  if (total <= 0) return 0;
  const target = rand() * total;
  let acc = 0;
  for (let i = 0; i < weights.length; i++) {
    acc += weights[i]!;
    if (acc >= target) return i;
  }
  return weights.length - 1;
}

// --- Crossover -----------------------------------------------------------
function twoPointCrossover(
  symbols: ReadonlyArray<string>,
  parent1: ReadonlyMap<string, string>,
  parent2: ReadonlyMap<string, string>,
  cut1: number,
  cut2: number,
): Map<string, string> {
  const child = new Map<string, string>();
  const lo = Math.min(cut1, cut2);
  const hi = Math.max(cut1, cut2);
  for (let i = 0; i < symbols.length; i++) {
    const s = symbols[i]!;
    const fromP2 = i >= lo && i < hi;
    const src = fromP2 ? parent2 : parent1;
    child.set(s, src.get(s) ?? (parent1.get(s) ?? ''));
  }
  return child;
}

// --- Mutation: swap-letters preserves multiset ---------------------------
function mutateBySwap(
  symbols: ReadonlyArray<string>,
  mapping: Map<string, string>,
  mutationRate: number,
  rand: () => number,
): Map<string, string> {
  const out = cloneMapping(mapping);
  const n = symbols.length;
  if (n < 2) return out;
  for (let i = 0; i < n; i++) {
    if (rand() < mutationRate) {
      let j = Math.floor(rand() * n);
      if (j === i) j = (j + 1) % n;
      const si = symbols[i]!;
      const sj = symbols[j]!;
      const a = out.get(si)!;
      const b = out.get(sj)!;
      out.set(si, b);
      out.set(sj, a);
    }
  }
  return out;
}

/**
 * Handles the gnosis solve Memetic Homophonic workflow.
 */
export function solveMemeticHomophonic(
  args: MemeticSolverArgs,
): MemeticSolverResult {
  // --- Validation ---
  if (!args.ciphertext || args.ciphertext.length === 0) {
    throw new RangeError('ciphertext must be non-empty');
  }
  const populationSize = args.populationSize ?? 24;
  if (populationSize < 4) {
    throw new RangeError('populationSize must be >= 4');
  }
  const generations = args.generations ?? 30;
  if (generations < 1) {
    throw new RangeError('generations must be >= 1');
  }
  const saSwapsPerIndividual = args.saSwapsPerIndividual ?? 5000;
  if (saSwapsPerIndividual < 100) {
    throw new RangeError('saSwapsPerIndividual must be >= 100');
  }
  const logisticR = args.logisticR ?? 3.99;
  if (logisticR <= 3.5699 || logisticR > 4) {
    throw new RangeError(
      'logisticR must be in (3.5699, 4] — chaotic regime required',
    );
  }
  const mutationRateBase = args.mutationRateBase ?? 0.10;
  const seed = args.seed ?? 0xC0DE_5EED;

  // Outer-loop fitness function. Default: combinedEnglishScore (smoother
  // gradient → better GA selection). Note: the inner SA polish via
  // solveHomophonicSubstitution always uses quadgramScore internally;
  // we re-score its results with `scoreFn` so GA ranking is consistent.
  const scoreFn = (args.useCombinedScore ?? true)
    ? combinedEnglishScore
    : quadgramScore;

  const ciphertext = args.ciphertext;

  // --- Step 2: initialize population via SA-forks (modest budget) -------
  const initBudget = Math.max(100, Math.floor(saSwapsPerIndividual / 2));
  const initResult: HomophonicSolverResult = solveHomophonicSubstitution({
    ciphertext,
    forks: populationSize,
    maxSwapsPerFork: initBudget,
    seed,
    strategy: 'simulated-annealing',
    annealing: {
      temperatureStart: args.saTemperatureStart,
      temperatureEnd: args.saTemperatureEnd,
      coolingRate: args.saCoolingRate,
    },
  });

  const symbols: ReadonlyArray<string> = initResult.symbols;
  if (symbols.length === 0) {
    throw new RangeError('ciphertext must contain at least one symbol');
  }

  // Build population from candidates. If there are fewer candidates than
  // populationSize (e.g. dedup), top up with copies of the best.
  // Re-score each candidate with `scoreFn` — the inner SA reports
  // quadgram-only scores, but GA ranking must use the chosen function.
  let population: Individual[] = [];
  const candidateList = initResult.candidates as ReadonlyArray<{
    mapping: ReadonlyMap<string, string>;
    score: number;
  }>;
  for (const c of candidateList) {
    population.push({
      mapping: cloneMapping(c.mapping),
      score: scoreMapping(ciphertext, c.mapping, scoreFn),
    });
    if (population.length >= populationSize) break;
  }
  while (population.length < populationSize) {
    const top = population[0] ?? {
      mapping: cloneMapping(initResult.mapping),
      score: scoreMapping(ciphertext, initResult.mapping, scoreFn),
    };
    population.push({ mapping: cloneMapping(top.mapping), score: top.score });
  }

  // Sort population by descending fitness.
  population.sort((a, b) => b.score - a.score);

  // Track best-ever.
  let bestEver: Individual = {
    mapping: cloneMapping(population[0]!.mapping),
    score: population[0]!.score,
  };

  const scoreHistory: number[] = [];
  const chaos = logisticMap(0.5, logisticR);
  const rand = makeLcg(seed ^ 0xA5A5_5A5A);

  // Per-individual SA seed stream so each polish run is deterministic.
  const seedStream = makeLcg(seed ^ 0x1234_5678);
  const nextSaSeed = (): number =>
    Math.floor(seedStream() * 0x7fff_ffff) >>> 0;

  // --- Generation loop --------------------------------------------------
  for (let g = 0; g < generations; g++) {
    // (a) Chaos-driven mutation rate.
    const x = chaos();
    const mutationRate = mutationRateBase * x;

    // (b) SA-polish each individual incrementally.
    const polishBudget = Math.max(100, Math.floor(saSwapsPerIndividual / 2));
    const polished: Individual[] = [];
    for (let i = 0; i < population.length; i++) {
      const ind = population[i]!;
      const polishedText = applyMapping(ciphertext, ind.mapping);
      // Run SA fork=1 starting from current text. We pass the polished
      // text as the ciphertext and an identity-like mapping, but the
      // homophonic solver works on the ciphertext alphabet directly,
      // so we just rerun on the original ciphertext with a fresh seed
      // and absorb whatever local optimum emerges. To respect the
      // "incremental polish" semantic, we fork with fewer swaps.
      void polishedText;
      const polishedRes = solveHomophonicSubstitution({
        ciphertext,
        forks: 1,
        maxSwapsPerFork: polishBudget,
        seed: nextSaSeed(),
        strategy: 'simulated-annealing',
        annealing: {
          temperatureStart: args.saTemperatureStart,
          temperatureEnd: args.saTemperatureEnd,
          coolingRate: args.saCoolingRate,
        },
      });
      // Pick whichever is better: the freshly-polished candidate or the
      // pre-existing individual. This guarantees monotone non-decrease
      // in best-of-population fitness across generations (within noise).
      // Re-score polishedRes with `scoreFn` — solveHomophonicSubstitution
      // returns scores in pure-quadgram terms; GA selection needs them
      // in the chosen (combined-or-quadgram) terms for consistency.
      const polishedScore = scoreMapping(ciphertext, polishedRes.mapping, scoreFn);
      if (polishedScore > ind.score) {
        polished.push({
          mapping: cloneMapping(polishedRes.mapping),
          score: polishedScore,
        });
      } else {
        polished.push({ mapping: cloneMapping(ind.mapping), score: ind.score });
      }
    }
    population = polished;

    // (c) Sort by descending fitness.
    population.sort((a, b) => b.score - a.score);

    // (d) Buleyean weights — rank-0 parent gets weight R+1, rank-(N-1)
    //     gets weight 2 (since w(R, R-1) = R - (R-1) + 1 = 2).
    const N = population.length;
    const weights: number[] = new Array(N);
    for (let r = 0; r < N; r++) {
      // rank r from top (best=0). Rejections = r. Rounds = N.
      weights[r] = buleyeanWeight(N, r);
    }

    // (f) Elitism: top quarter preserved unchanged.
    const eliteCount = Math.max(1, Math.floor(N / 4));
    const elites = population.slice(0, eliteCount).map((ind) => ({
      mapping: cloneMapping(ind.mapping),
      score: ind.score,
    }));

    // (e) Generate children to replace the bottom half.
    const halfPoint = Math.floor(N / 2);
    const nextPop: Individual[] = population.slice(0, halfPoint).map((ind) => ({
      mapping: cloneMapping(ind.mapping),
      score: ind.score,
    }));

    while (nextPop.length < N) {
      // Pick parents.
      const p1Idx = pickIndexByWeight(weights, rand);
      let p2Idx = pickIndexByWeight(weights, rand);
      if (p2Idx === p1Idx) p2Idx = (p2Idx + 1) % N;
      const parent1 = population[p1Idx]!;
      const parent2 = population[p2Idx]!;

      // Cut points from chaos iterator.
      const c1 = Math.floor(chaos() * symbols.length);
      const c2 = Math.floor(chaos() * symbols.length);

      const childMapping = twoPointCrossover(
        symbols,
        parent1.mapping,
        parent2.mapping,
        c1,
        c2,
      );

      const mutated = mutateBySwap(symbols, childMapping, mutationRate, rand);
      const childScore = scoreMapping(ciphertext, mutated, scoreFn);
      nextPop.push({ mapping: mutated, score: childScore });
    }

    // Apply elitism: ensure top eliteCount are the elites (replace the
    // worst eliteCount of the new population with the elites).
    nextPop.sort((a, b) => b.score - a.score);
    for (let i = 0; i < eliteCount; i++) {
      // Replace the i-th-from-bottom slot with elite i.
      const targetIdx = nextPop.length - 1 - i;
      if (elites[i]!.score > nextPop[targetIdx]!.score) {
        nextPop[targetIdx] = elites[i]!;
      }
    }
    nextPop.sort((a, b) => b.score - a.score);

    population = nextPop;

    // (g) Update best-ever.
    const currentBest = population[0]!;
    if (currentBest.score > bestEver.score) {
      bestEver = {
        mapping: cloneMapping(currentBest.mapping),
        score: currentBest.score,
      };
    }

    // Stats.
    const scores = population.map((ind) => ind.score);
    const bestScore = scores[0]!;
    const worstScore = scores[scores.length - 1]!;
    const meanScore = meanOf(scores);
    const diversity = stddevOf(scores);
    scoreHistory.push(bestEver.score);

    // (h) Callback.
    if (args.onGeneration) {
      args.onGeneration({
        generation: g,
        bestScore,
        worstScore,
        meanScore,
        diversity,
        mutationRate,
        bestPlaintext: applyMapping(ciphertext, bestEver.mapping),
      });
    }
  }

  // Final stats from the last population.
  const finalScores = population.map((ind) => ind.score);
  const finalBest = finalScores[0] ?? bestEver.score;
  const finalWorst = finalScores[finalScores.length - 1] ?? bestEver.score;
  const finalMean = meanOf(finalScores);

  return {
    plaintext: applyMapping(ciphertext, bestEver.mapping),
    mapping: bestEver.mapping,
    score: bestEver.score,
    generations,
    finalPopulationStats: {
      bestScore: finalBest,
      worstScore: finalWorst,
      meanScore: finalMean,
    },
    scoreHistory,
    symbols,
  };
}

// =========================================================================
// Parallel variant — solveMemeticHomophonicParallel
// =========================================================================
// Worker-pool implementation that mirrors the sequential `solveMemeticHomophonic`
// generation loop but dispatches the per-individual SA polish step across
// worker_threads. Each generation issues `population.length` polish tasks to
// the pool, awaits all of them via Promise.all, then runs GA selection +
// crossover + mutation on the main thread (these steps are cheap relative
// to the SA polish).
//
// API: distinct exported function. Existing `solveMemeticHomophonic` retains
// its synchronous signature; callers wanting parallelism opt in by switching
// to this function (and awaiting). Same args type plus the `parallel` field.
//
// Worker lifecycle: one Worker per pool slot, kept alive across generations,
// shut down via { type: 'shutdown' } in a finally block. Each generation
// uses round-robin dispatch with a queue when populationSize > workerCount.

interface PolishTaskMessage {
  type: 'polish';
  taskId: number;
  ciphertext: string;
  seed: number;
  saParams: {
    maxSwapsPerFork: number;
    temperatureStart?: number;
    temperatureEnd?: number;
    coolingRate?: number;
  };
}

interface PolishDoneMessage {
  type: 'done';
  taskId: number;
  mapping: Array<[string, string]>;
  score: number;
}

interface PolishErrorMessage {
  type: 'error';
  taskId: number;
  error: string;
}

type PolishReplyMessage = PolishDoneMessage | PolishErrorMessage;

interface DispatchResult {
  taskId: number;
  mapping: Map<string, string>;
}

/**
 * Dispatch one polish task and resolve when the matching reply arrives.
 * Resolves the worker back to the idle pool via the supplied `release`
 * callback so the caller can reuse it for the next queued task.
 */
function dispatchPolishOnWorker(
  worker: Worker,
  task: PolishTaskMessage,
  release: () => void,
): Promise<DispatchResult> {
  return new Promise<DispatchResult>((resolve, reject) => {
    const onMessage = (msg: PolishReplyMessage | { type: string }): void => {
      if (!msg || (msg as { type: string }).type === 'ready') return;
      const reply = msg as PolishReplyMessage;
      if (reply.taskId !== task.taskId) return;
      worker.off('message', onMessage);
      worker.off('error', onError);
      release();
      if (reply.type === 'done') {
        resolve({
          taskId: reply.taskId,
          mapping: new Map(reply.mapping),
        });
      } else {
        reject(new Error(`worker task ${reply.taskId} failed: ${reply.error}`));
      }
    };
    const onError = (err: Error): void => {
      worker.off('message', onMessage);
      worker.off('error', onError);
      release();
      reject(err);
    };
    worker.on('message', onMessage);
    worker.on('error', onError);
    worker.postMessage(task);
  });
}

/**
 * Handles the gnosis solve Memetic Homophonic Parallel workflow.
 */
export async function solveMemeticHomophonicParallel(
  args: MemeticSolverArgs,
): Promise<MemeticSolverResult> {
  // --- Validation (mirrors sequential path) ---
  if (!args.ciphertext || args.ciphertext.length === 0) {
    throw new RangeError('ciphertext must be non-empty');
  }
  const populationSize = args.populationSize ?? 24;
  if (populationSize < 4) {
    throw new RangeError('populationSize must be >= 4');
  }
  const generations = args.generations ?? 30;
  if (generations < 1) {
    throw new RangeError('generations must be >= 1');
  }
  const saSwapsPerIndividual = args.saSwapsPerIndividual ?? 5000;
  if (saSwapsPerIndividual < 100) {
    throw new RangeError('saSwapsPerIndividual must be >= 100');
  }
  const logisticR = args.logisticR ?? 3.99;
  if (logisticR <= 3.5699 || logisticR > 4) {
    throw new RangeError(
      'logisticR must be in (3.5699, 4] — chaotic regime required',
    );
  }
  const mutationRateBase = args.mutationRateBase ?? 0.10;
  const seed = args.seed ?? 0xC0DE_5EED;

  const scoreFn = (args.useCombinedScore ?? true)
    ? combinedEnglishScore
    : quadgramScore;

  const ciphertext = args.ciphertext;

  // --- Spin up the worker pool ------------------------------------------
  const workerCount =
    args.parallel?.workers ?? Math.max(1, Math.floor(os.cpus().length / 2));
  const here = dirname(fileURLToPath(import.meta.url));
  const workerPath = resolvePath(here, './memetic-worker.ts');
  const pool: Worker[] = [];
  const idleQueue: Worker[] = [];
  const waiters: Array<(w: Worker) => void> = [];

  const release = (w: Worker): void => {
    const next = waiters.shift();
    if (next) next(w);
    else idleQueue.push(w);
  };
  const acquire = (): Promise<Worker> => {
    const w = idleQueue.shift();
    if (w) return Promise.resolve(w);
    return new Promise<Worker>((resolve) => {
      waiters.push(resolve);
    });
  };

  for (let i = 0; i < workerCount; i++) {
    // Inherit the parent's execArgv so the tsx ESM loader (preflight.cjs +
    // loader.mjs) propagates into the worker. `['--import', 'tsx']` alone
    // doesn't carry the loader URL the parent was launched with.
    const w = new Worker(workerPath, {
      execArgv: process.execArgv,
    });
    pool.push(w);
    // Wait for the worker's initial 'ready' handshake before adding to
    // the idle queue — guards against a task being posted before the
    // worker's message handler is wired up inside the worker process.
    await new Promise<void>((resolve, reject) => {
      const onReady = (msg: { type?: string }): void => {
        if (msg && msg.type === 'ready') {
          w.off('message', onReady);
          w.off('error', onErr);
          idleQueue.push(w);
          resolve();
        }
      };
      const onErr = (err: Error): void => {
        w.off('message', onReady);
        w.off('error', onErr);
        reject(err);
      };
      w.on('message', onReady);
      w.on('error', onErr);
    });
  }

  try {
    // --- Step 2: initialize population via SA-forks (modest budget) -----
    // Done on main thread — populationSize forks in one call is already
    // an internal FRF parallel by the homophonic solver; replicating it
    // across workers would be churn for a one-shot init step.
    const initBudget = Math.max(100, Math.floor(saSwapsPerIndividual / 2));
    const initResult: HomophonicSolverResult = solveHomophonicSubstitution({
      ciphertext,
      forks: populationSize,
      maxSwapsPerFork: initBudget,
      seed,
      strategy: 'simulated-annealing',
      annealing: {
        temperatureStart: args.saTemperatureStart,
        temperatureEnd: args.saTemperatureEnd,
        coolingRate: args.saCoolingRate,
      },
    });

    const symbols: ReadonlyArray<string> = initResult.symbols;
    if (symbols.length === 0) {
      throw new RangeError('ciphertext must contain at least one symbol');
    }

    let population: Individual[] = [];
    const candidateList = initResult.candidates as ReadonlyArray<{
      mapping: ReadonlyMap<string, string>;
      score: number;
    }>;
    for (const c of candidateList) {
      population.push({
        mapping: cloneMapping(c.mapping),
        score: scoreMapping(ciphertext, c.mapping, scoreFn),
      });
      if (population.length >= populationSize) break;
    }
    while (population.length < populationSize) {
      const top = population[0] ?? {
        mapping: cloneMapping(initResult.mapping),
        score: scoreMapping(ciphertext, initResult.mapping, scoreFn),
      };
      population.push({ mapping: cloneMapping(top.mapping), score: top.score });
    }

    population.sort((a, b) => b.score - a.score);

    let bestEver: Individual = {
      mapping: cloneMapping(population[0]!.mapping),
      score: population[0]!.score,
    };

    const scoreHistory: number[] = [];
    const chaos = logisticMap(0.5, logisticR);
    const rand = makeLcg(seed ^ 0xA5A5_5A5A);
    const seedStream = makeLcg(seed ^ 0x1234_5678);
    const nextSaSeed = (): number =>
      Math.floor(seedStream() * 0x7fff_ffff) >>> 0;

    // --- Generation loop ------------------------------------------------
    for (let g = 0; g < generations; g++) {
      const x = chaos();
      const mutationRate = mutationRateBase * x;

      // (b) Parallel SA-polish: dispatch one task per individual.
      const polishBudget = Math.max(100, Math.floor(saSwapsPerIndividual / 2));
      const saParams = {
        maxSwapsPerFork: polishBudget,
        temperatureStart: args.saTemperatureStart,
        temperatureEnd: args.saTemperatureEnd,
        coolingRate: args.saCoolingRate,
      };

      // Snapshot the pre-polish population so we can compare each polished
      // result against the original individual (matches sequential
      // semantics: keep whichever is better).
      const preIndividuals = population.map((ind) => ({
        mapping: cloneMapping(ind.mapping),
        score: ind.score,
      }));

      const polishPromises: Array<Promise<Individual>> = preIndividuals.map(
        (ind, idx) => {
          const task: PolishTaskMessage = {
            type: 'polish',
            taskId: g * populationSize + idx,
            ciphertext,
            seed: nextSaSeed(),
            saParams,
          };
          return acquire().then(async (w) => {
            const reply = await dispatchPolishOnWorker(w, task, () =>
              release(w),
            );
            const polishedScore = scoreMapping(
              ciphertext,
              reply.mapping,
              scoreFn,
            );
            if (polishedScore > ind.score) {
              return {
                mapping: reply.mapping,
                score: polishedScore,
              };
            }
            return {
              mapping: cloneMapping(ind.mapping),
              score: ind.score,
            };
          });
        },
      );

      population = await Promise.all(polishPromises);

      // (c) Sort by descending fitness.
      population.sort((a, b) => b.score - a.score);

      // (d) Buleyean weights.
      const N = population.length;
      const weights: number[] = new Array(N);
      for (let r = 0; r < N; r++) {
        weights[r] = buleyeanWeight(N, r);
      }

      // (f) Elitism: top quarter preserved unchanged.
      const eliteCount = Math.max(1, Math.floor(N / 4));
      const elites = population.slice(0, eliteCount).map((ind) => ({
        mapping: cloneMapping(ind.mapping),
        score: ind.score,
      }));

      // (e) Generate children to replace the bottom half.
      const halfPoint = Math.floor(N / 2);
      const nextPop: Individual[] = population.slice(0, halfPoint).map(
        (ind) => ({
          mapping: cloneMapping(ind.mapping),
          score: ind.score,
        }),
      );

      while (nextPop.length < N) {
        const p1Idx = pickIndexByWeight(weights, rand);
        let p2Idx = pickIndexByWeight(weights, rand);
        if (p2Idx === p1Idx) p2Idx = (p2Idx + 1) % N;
        const parent1 = population[p1Idx]!;
        const parent2 = population[p2Idx]!;

        const c1 = Math.floor(chaos() * symbols.length);
        const c2 = Math.floor(chaos() * symbols.length);

        const childMapping = twoPointCrossover(
          symbols,
          parent1.mapping,
          parent2.mapping,
          c1,
          c2,
        );

        const mutated = mutateBySwap(symbols, childMapping, mutationRate, rand);
        const childScore = scoreMapping(ciphertext, mutated, scoreFn);
        nextPop.push({ mapping: mutated, score: childScore });
      }

      nextPop.sort((a, b) => b.score - a.score);
      for (let i = 0; i < eliteCount; i++) {
        const targetIdx = nextPop.length - 1 - i;
        if (elites[i]!.score > nextPop[targetIdx]!.score) {
          nextPop[targetIdx] = elites[i]!;
        }
      }
      nextPop.sort((a, b) => b.score - a.score);

      population = nextPop;

      const currentBest = population[0]!;
      if (currentBest.score > bestEver.score) {
        bestEver = {
          mapping: cloneMapping(currentBest.mapping),
          score: currentBest.score,
        };
      }

      const scores = population.map((ind) => ind.score);
      const bestScore = scores[0]!;
      const worstScore = scores[scores.length - 1]!;
      const meanScore = meanOf(scores);
      const diversity = stddevOf(scores);
      scoreHistory.push(bestEver.score);

      if (args.onGeneration) {
        args.onGeneration({
          generation: g,
          bestScore,
          worstScore,
          meanScore,
          diversity,
          mutationRate,
          bestPlaintext: applyMapping(ciphertext, bestEver.mapping),
        });
      }
    }

    const finalScores = population.map((ind) => ind.score);
    const finalBest = finalScores[0] ?? bestEver.score;
    const finalWorst = finalScores[finalScores.length - 1] ?? bestEver.score;
    const finalMean = meanOf(finalScores);

    return {
      plaintext: applyMapping(ciphertext, bestEver.mapping),
      mapping: bestEver.mapping,
      score: bestEver.score,
      generations,
      finalPopulationStats: {
        bestScore: finalBest,
        worstScore: finalWorst,
        meanScore: finalMean,
      },
      scoreHistory,
      symbols,
    };
  } finally {
    // Cleanup: shut down all workers. Wrapped in try-blocks so a single
    // worker that's already exited doesn't prevent the rest from being
    // terminated.
    for (const w of pool) {
      try {
        w.postMessage({ type: 'shutdown' });
      } catch {
        // worker may already be dead — ignore.
      }
    }
    await Promise.all(
      pool.map((w) =>
        w.terminate().catch(() => {
          /* ignore termination errors */
        }),
      ),
    );
  }
}
