// @a0n/gnosis/distributed-inference-host/src/memo-client.ts
/**
 * memo-client.ts — TS client for the `/memo/*` HTTP knob set.
 *
 * Wire-compatible with both:
 *   - The wasm worker  (`apps/distributed-inference-worker/src/index.ts`)
 *   - The native bin   (`open-source/gnosis/distributed-inference/src/bin/fat-station-memo.rs`)
 *
 * Both surfaces are byte-identical: same JSON shapes, same packed-f32 LE
 * body for `/memo/probe-drift`. This module gives the host (chain.ts /
 * pipeline.ts) one place to flip the ER=EPR teleportation cache on/off
 * per session and to read out hit-rate + Pisot-drift telemetry.
 *
 * ARCHITECTURAL STATUS
 * --------------------
 *   - Wired, hot-path-ready against fat-station-memo (smoke-validated by
 *     scripts/probe-memo-knob.sh: 9/9 routes pass).
 *   - Not yet called from chain.ts — that's the integration step. The
 *     intended call site is the per-session boot path in
 *     `pipeline_pneuma.ts` / `pipeline.ts`: configure once with the
 *     bench-validated tau² (1e-4), then poll stats periodically to feed
 *     the adaptive router (when it lands).
 *
 * USAGE
 * -----
 *   const memo = new MemoClient({ baseUrl: 'http://127.0.0.1:8765' });
 *   await memo.configure({ enabled: true, tauSquared: 1e-4, maxEntries: 8192 });
 *   const stats = await memo.stats();        // { hits, misses, evictions, ... }
 *   const drift = await memo.drift(0.5);     // luminary-fraction histogram
 *   const probe = await memo.probeDrift(new Float32Array([...]));
 *   await memo.clear();
 *
 * Or against an `IStation`-like endpoint (service binding, test fetcher):
 *   const memo = new MemoClient({ endpoint: { fetch }, baseUrl: 'http://station' });
 *
 * See `docs/pneuma/memo-routes.md` for the full route table and
 * recommended production settings.
 */

export interface MemoFetcher {
  fetch(req: Request): Promise<Response>;
}

type MemoPostBody = string | Uint8Array | ArrayBuffer;

export interface MemoClientConfig {
  /** Either a base URL (HTTP) or `{ endpoint, baseUrl? }` (service-binding). */
  baseUrl?: string;
  endpoint?: MemoFetcher;
}

export interface MemoStats {
  hits: number;
  misses: number;
  evictions: number;
  total: number;
  hitRate: number;
}

export interface MemoConfigure {
  enabled?: boolean;
  /** L²-distance² hit threshold. Bench-validated production value: 1e-4. */
  tauSquared?: number;
  /** Eviction cap. Default 4096 (matches `MatVecMemo::new`). */
  maxEntries?: number;
}

export interface MemoDriftSummary {
  luminary: number;
  total: number;
  finite: number;
  nanCount: number;
  luminaryFraction: number;
  luminaryThreshold: number;
  driftMin: number;
  driftMean: number;
  driftMax: number;
  /** Surrogate's real-valued minimum (≈ 0.345 at ‖x‖ = φ/2). */
  surrogateMinimum: number;
  note?: string;
}

export interface MemoProbeDrift {
  /** null when the surrogate evaluates to NaN/Inf (broken kernel input). */
  drift: number | null;
  driftFinite: boolean;
  norm: number;
  normSquared: number;
  length: number;
  surrogateMinimum: number;
  formula: string;
}

export interface MemoSelftestStats {
  hits: number;
  misses: number;
  evictions: number;
}

export interface MemoSelftestResult {
  cold_ms: number[];
  warm1_ms: number[];
  warm2_ms: number[];
  stats_after_fill: MemoSelftestStats;
  stats_final: MemoSelftestStats;
  drift: {
    luminary: number;
    total: number;
    finite: number;
    drift_min: number;
    drift_mean: number;
    drift_max: number;
    luminary_threshold: number;
  };
  hidden_dim: number;
  steps: number;
  tau_squared: number;
  /** Native fat-station-memo includes this; the worker omits it. */
  layer?: number;
}

export class MemoClient {
  private readonly fetcher: MemoFetcher;
  private readonly baseUrl: string;

  constructor(config: MemoClientConfig) {
    if (!config.baseUrl && !config.endpoint) {
      throw new Error('MemoClient: baseUrl or endpoint is required');
    }
    this.baseUrl = (config.baseUrl ?? 'http://station').replace(/\/+$/, '');
    this.fetcher = config.endpoint ?? { fetch: (req) => fetch(req) };
  }

  /**
   * Enable the cache and tune τ² + capacity. Any field omitted from
   * `body` is left unchanged on the server.
   *
   * Bench-validated production setting: `{ enabled: true, tauSquared: 1e-4 }`.
   * At τ² = 1e-4 the cache reproduces correct logits (cosine ≥ 0.999 vs
   * baseline) while reducing wall-time by ~86% on the qwen-coder-7b
   * split-phase forward (see `bin/bench-teleportation.rs` output).
   */
  async configure(body: MemoConfigure): Promise<{ configured: true; stats: MemoSelftestStats }> {
    const wireBody: Record<string, number | boolean> = {};
    if (typeof body.enabled === 'boolean') wireBody.enabled = body.enabled;
    if (typeof body.tauSquared === 'number') wireBody.tau_squared = body.tauSquared;
    if (typeof body.maxEntries === 'number') wireBody.max_entries = body.maxEntries;
    const res = await this.post('/memo/configure', JSON.stringify(wireBody), 'application/json');
    return (await res.json()) as { configured: true; stats: MemoSelftestStats };
  }

  async stats(): Promise<MemoStats> {
    const res = await this.get('/memo/stats');
    const json = (await res.json()) as {
      hits: number;
      misses: number;
      evictions: number;
      total: number;
      hit_rate: number;
    };
    return {
      hits: json.hits,
      misses: json.misses,
      evictions: json.evictions,
      total: json.total,
      hitRate: json.hit_rate,
    };
  }

  /**
   * Pisot-drift histogram across cached entries.
   *
   * `threshold` is the f32-domain luminary cutoff. Default 0.5 — just
   * above the surrogate minimum (≈ 0.345). Pass NaN/null to fall back
   * to the server default. See the wasm worker handler doc
   * (`apps/distributed-inference-worker/src/index.ts:1046`) for the
   * scope caveat: this is the f32 surrogate, not the Lean witness on
   * integer Lucas pairs.
   */
  async drift(threshold = 0.5): Promise<MemoDriftSummary> {
    const url = `/memo/drift?threshold=${encodeURIComponent(String(threshold))}`;
    const res = await this.get(url);
    const json = (await res.json()) as {
      luminary: number;
      total: number;
      finite: number;
      nan_count: number;
      luminary_fraction: number;
      luminary_threshold: number;
      drift_min: number;
      drift_mean: number;
      drift_max: number;
      surrogate_minimum: number;
      note?: string;
    };
    return {
      luminary: json.luminary,
      total: json.total,
      finite: json.finite,
      nanCount: json.nan_count,
      luminaryFraction: json.luminary_fraction,
      luminaryThreshold: json.luminary_threshold,
      driftMin: json.drift_min,
      driftMean: json.drift_mean,
      driftMax: json.drift_max,
      surrogateMinimum: json.surrogate_minimum,
      note: json.note,
    };
  }

  /**
   * Compute the Pisot-drift surrogate on an arbitrary f32 vector
   * WITHOUT inserting into the cache. Useful as a sanity check
   * independent of the matvec kernel chain — drift on a healthy random
   * residual should be a finite positive number near the surrogate
   * minimum (≈ 0.345 for ‖x‖ near φ/2).
   */
  async probeDrift(x: Float32Array): Promise<MemoProbeDrift> {
    if (x.length === 0) {
      throw new Error('MemoClient.probeDrift: input must be non-empty');
    }
    // Build a fresh ArrayBuffer view so we send ONLY the typed-array
    // bytes (avoids accidental over-send when the Float32Array is a
    // view onto a larger backing buffer).
    const bytes = new Uint8Array(x.buffer.slice(x.byteOffset, x.byteOffset + x.byteLength));
    const res = await this.post('/memo/probe-drift', bytes, 'application/octet-stream');
    const json = (await res.json()) as {
      drift: number | null;
      drift_finite: boolean;
      norm: number;
      norm_squared: number;
      length: number;
      surrogate_minimum: number;
      formula: string;
    };
    return {
      drift: json.drift,
      driftFinite: json.drift_finite,
      norm: json.norm,
      normSquared: json.norm_squared,
      length: json.length,
      surrogateMinimum: json.surrogate_minimum,
      formula: json.formula,
    };
  }

  async clear(): Promise<{ cleared: true }> {
    const res = await this.post('/memo/clear', '', 'application/json');
    return (await res.json()) as { cleared: true };
  }

  /**
   * In-process cold/warm bench. Drives the full attn→ffn split-phase
   * sequence `steps` times under three conditions (cold, warm-fill,
   * warm-rerun) so cross-isolate jitter doesn't lie about cache wins.
   *
   * Requires the served station to own layer 0 (entry|both). Stations
   * that don't will 404 — call against the entry station only.
   */
  async selftest(opts: {
    steps?: number;
    /** Default = the station's hidden_dim. Mismatched values are silently coerced. */
    hiddenDim?: number;
    tauSquared?: number;
    maxEntries?: number;
    /** Native fat-station-memo only: which layer to bench against. */
    layer?: number;
  } = {}): Promise<MemoSelftestResult> {
    const wire: Record<string, number> = {};
    if (typeof opts.steps === 'number') wire.steps = opts.steps;
    if (typeof opts.hiddenDim === 'number') wire.hidden_dim = opts.hiddenDim;
    if (typeof opts.tauSquared === 'number') wire.tau_squared = opts.tauSquared;
    if (typeof opts.maxEntries === 'number') wire.max_entries = opts.maxEntries;
    if (typeof opts.layer === 'number') wire.layer = opts.layer;
    const res = await this.post('/memo/selftest', JSON.stringify(wire), 'application/json');
    return (await res.json()) as MemoSelftestResult;
  }

  // ── transport helpers ───────────────────────────────────────────────
  private async get(path: string): Promise<Response> {
    const res = await this.fetcher.fetch(new Request(`${this.baseUrl}${path}`));
    if (!res.ok) throw new Error(`MemoClient GET ${path}: ${res.status} ${await safeText(res)}`);
    return res;
  }

  private async post(path: string, body: MemoPostBody, contentType: string): Promise<Response> {
    const res = await this.fetcher.fetch(
      new Request(`${this.baseUrl}${path}`, {
        method: 'POST',
        headers: { 'Content-Type': contentType },
        body: body as any,
      }),
    );
    if (!res.ok) throw new Error(`MemoClient POST ${path}: ${res.status} ${await safeText(res)}`);
    return res;
  }
}

async function safeText(res: Response): Promise<string> {
  try {
    return (await res.text()).slice(0, 500);
  } catch {
    return '<no body>';
  }
}
