// @a0n/gnosis/distributed-inference-host/src/mesh-carrier-assessment.ts
import { KARMIC_MESH_RESIDUAL_VANISHES_THEOREM } from '@a0n/aether/src/karmic-attention-optimizer';

/**
 * Stable JSON shape for mesh span / response-header carrier telemetry:
 * spectral (Bule + fingerprint) governance, optional attention-closure lift
 * lineage, optional deblur FRF fold summary, first-class epistemic outcome
 * (admit / withhold / observation-only). Same envelope for D1 rows and
 * `X-Mesh-Carrier-Assessment` on octet-stream responses.
 */
import type { DeblurEnsembleResult } from './forkraceFoldDeblur';
import {
  computeHornObservabilityFromSpectralAssessment,
  type HornObservabilityRuntimeWire,
} from './horn-observability-runtime';
import type { InvisibleWatermarkProvenance } from './invisible-watermark';
import type {
  MeshAttentionClosureAssessment,
  SpectralNoiseAssessment,
} from './spectral-noise-monitor';
import type { StyleTransferEnsembleResult } from './style-transfer-types';

/** Request header: base64(JSON) `deblurFold` wire merged into `mesh_carrier_assessment/v1` on the worker. */
export const MESH_HTTP_HEADER_CARRIER_DEBLUR_FOLD =
  'X-Mesh-Carrier-Deblur-Fold' as const;

/** Request header: base64(JSON) style-transfer FRF fold wire (same merge pattern as deblur). */
export const MESH_HTTP_HEADER_CARRIER_STYLE_FOLD =
  'X-Mesh-Carrier-Style-Fold' as const;

export const MESH_CARRIER_ASSESSMENT_CERTIFICATE_VERSION =
  'mesh_carrier_assessment/v1' as const;

/** Typed mesh hop claim posture (same certificate as spectral telemetry). */
export const MESH_CARRIER_EPISTEMIC_OUTCOMES = [
  'claim_admitted',
  'claim_withheld',
  'observation_only',
] as const;
export type MeshCarrierEpistemicOutcome =
  (typeof MESH_CARRIER_EPISTEMIC_OUTCOMES)[number];

/** Which limitation class constrained the hop (non-empty semantics for ops). */
export const MESH_CARRIER_EPISTEMIC_LIMITATION_CLASSES = [
  'global_optimality_unavailable',
  'spectral_boundary_pressure',
  'governance_breach',
  'closure_lift_pressure',
  'compute_enumeration_bound',
  'void_no_answering_reason',
  'other',
] as const;
export type MeshCarrierEpistemicLimitationClass =
  (typeof MESH_CARRIER_EPISTEMIC_LIMITATION_CLASSES)[number];

export interface MeshCarrierEpistemicWire {
  outcome: MeshCarrierEpistemicOutcome;
  limitationClasses: readonly MeshCarrierEpistemicLimitationClass[];
  summary: string;
  theoremLineage?: readonly string[];
}

const EPISTEMIC_LIMITATION_ORDER: readonly MeshCarrierEpistemicLimitationClass[] =
  MESH_CARRIER_EPISTEMIC_LIMITATION_CLASSES;

/**
 * Derives a deterministic `epistemic` block from spectral + optional attention
 * closure (typed refusal / pressure vs nominal admission).
 */
export function deriveMeshCarrierEpistemicFromSpectral(
  spectral: SpectralNoiseAssessment,
): MeshCarrierEpistemicWire {
  const limits = new Set<MeshCarrierEpistemicLimitationClass>();
  limits.add('global_optimality_unavailable');
  if (spectral.governanceViolation !== null) {
    limits.add('governance_breach');
  }
  if (
    spectral.state === 'pre_collapse' ||
    spectral.state === 'boundary_collapse'
  ) {
    limits.add('spectral_boundary_pressure');
  }
  if (spectral.action !== 'none') {
    limits.add('spectral_boundary_pressure');
  }
  const ac = spectral.attentionClosure;
  if (ac !== undefined && ac.decision.action === 'lift_resolution') {
    limits.add('closure_lift_pressure');
  }

  let outcome: MeshCarrierEpistemicOutcome;
  if (
    spectral.state === 'boundary_collapse' ||
    spectral.governanceViolation !== null
  ) {
    outcome = 'claim_withheld';
  } else if (spectral.state === 'pre_collapse' || spectral.action !== 'none') {
    outcome = 'claim_withheld';
  } else {
    outcome = 'claim_admitted';
  }

  const limitationClasses = EPISTEMIC_LIMITATION_ORDER.filter((c) =>
    limits.has(c),
  );
  let theoremLineage: readonly string[] | undefined =
    ac !== undefined ? [...ac.workPolicy.theoremLineage] : undefined;
  if (
    theoremLineage !== undefined &&
    ac !== undefined &&
    ac.workPolicy.allowHeadPruning
  ) {
    theoremLineage = [...theoremLineage, KARMIC_MESH_RESIDUAL_VANISHES_THEOREM];
  }

  return {
    outcome,
    limitationClasses,
    summary:
      spectral.reason.trim().length > 0
        ? spectral.reason
        : 'spectral assessment without reason string',
    theoremLineage,
  };
}

/** Optional summary when a caller attaches FRF/deblur diagnostics to the hop. */
export interface MeshCarrierDeblurFoldWire {
  foldRounds: number;
  foldSurvivors: number;
  foldRejected: number;
  foldTotalWeight: number;
  candidateLabels: readonly string[];
  selectedSigmas: readonly number[];
  watermark?: InvisibleWatermarkProvenance;
}

/** Optional summary when a caller attaches `forkRaceFoldStyleTransfer` diagnostics. */
export interface MeshCarrierStyleTransferFoldWire {
  foldRounds: number;
  foldSurvivors: number;
  foldRejected: number;
  foldTotalWeight: number;
  candidateLabels: readonly string[];
  compositeScores: readonly number[];
  meanEffectiveGates: readonly number[];
}

export interface MeshCarrierAssessmentV1Wire {
  certificateVersion: typeof MESH_CARRIER_ASSESSMENT_CERTIFICATE_VERSION;
  stationLayerStart: number;
  stationLayerEndExclusive: number;
  spectral: {
    readonly state: SpectralNoiseAssessment['state'];
    readonly action: SpectralNoiseAssessment['action'];
    readonly fingerprint: SpectralNoiseAssessment['fingerprint'];
    readonly fingerprintBuley: SpectralNoiseAssessment['fingerprintBuley'];
    readonly traceBuley: SpectralNoiseAssessment['buley'];
    readonly governanceViolation: SpectralNoiseAssessment['governanceViolation'];
    readonly reason: string;
    readonly traceCompact: {
      readonly requestId: string;
      readonly nodeIdx: number;
      readonly chunkIdx: number;
      readonly startPos: number;
      readonly sequenceLength: number;
      readonly latencyMs: number;
      readonly requestEntropy: number;
      readonly boundaryLimit: number;
    };
  };
  attentionClosure?: {
    readonly meshAction: MeshAttentionClosureAssessment['action'];
    readonly decisionAction: MeshAttentionClosureAssessment['decision']['action'];
    readonly decisionReason: string;
    readonly telemetryLabel: MeshAttentionClosureAssessment['workPolicy']['telemetryLabel'];
    readonly theoremLineage: readonly string[];
    readonly preserveHeads: boolean;
    readonly observer: {
      readonly heads: number;
      readonly coupling: number;
      readonly observerResolution: number;
      readonly kind: MeshAttentionClosureAssessment['observed']['kind'];
      readonly unresolvedResidue: number;
      readonly traceSeparated: boolean;
    };
  };
  deblurFold?: MeshCarrierDeblurFoldWire;
  styleTransferFold?: MeshCarrierStyleTransferFoldWire;
  hornObservability?: HornObservabilityRuntimeWire;
  epistemic: MeshCarrierEpistemicWire;
}

export type MeshCarrierNoiseSignalInterpretation =
  | 'signal-supported'
  | 'noise-burdened';

export interface MeshCarrierNoiseSignalProjection {
  readonly spansProcessed: number;
  readonly certificatesObserved: number;
  readonly certificateCoverage: number;
  readonly readySignal: number;
  readonly recoverableSignal: number;
  readonly signal: number;
  readonly notReadyNoise: number;
  readonly partialObservationNoise: number;
  readonly noise: number;
  readonly noiseSignalLimit: number;
  readonly readinessMargin: number | null;
  readonly residuePressure: number | null;
  readonly deficitPressure: number | null;
  readonly saturated: boolean;
  readonly interpretation: MeshCarrierNoiseSignalInterpretation;
  readonly theoremLineage?: readonly string[];
}

export interface EncodeMeshCarrierAssessmentV1Args {
  stationLayerStart: number;
  stationLayerEndExclusive: number;
  spectral: SpectralNoiseAssessment;
  deblurFold?: MeshCarrierDeblurFoldWire | null;
  styleTransferFold?: MeshCarrierStyleTransferFoldWire | null;
}

/**
 * Converts input into Mesh Carrier Assessment Wire.
 */
export function toMeshCarrierAssessmentWire(
  args: EncodeMeshCarrierAssessmentV1Args,
): MeshCarrierAssessmentV1Wire {
  const { spectral } = args;
  const wire: MeshCarrierAssessmentV1Wire = {
    certificateVersion: MESH_CARRIER_ASSESSMENT_CERTIFICATE_VERSION,
    stationLayerStart: args.stationLayerStart,
    stationLayerEndExclusive: args.stationLayerEndExclusive,
    spectral: {
      state: spectral.state,
      action: spectral.action,
      fingerprint: spectral.fingerprint,
      fingerprintBuley: spectral.fingerprintBuley,
      traceBuley: spectral.buley,
      governanceViolation: spectral.governanceViolation,
      reason: spectral.reason,
      traceCompact: {
        requestId: spectral.trace.requestId,
        nodeIdx: spectral.trace.nodeIdx,
        chunkIdx: spectral.trace.chunkIdx,
        startPos: spectral.trace.startPos,
        sequenceLength: spectral.trace.sequenceLength,
        latencyMs: spectral.trace.latencyMs,
        requestEntropy: spectral.trace.requestEntropy,
        boundaryLimit: spectral.trace.boundaryLimit,
      },
    },
    hornObservability: computeHornObservabilityFromSpectralAssessment(spectral),
    epistemic: deriveMeshCarrierEpistemicFromSpectral(spectral),
  };
  const ac = spectral.attentionClosure;
  if (ac !== undefined) {
    wire.attentionClosure = attentionClosureAssessmentToWire(ac);
  }
  if (args.deblurFold !== undefined && args.deblurFold !== null) {
    wire.deblurFold = { ...args.deblurFold, candidateLabels: [...args.deblurFold.candidateLabels], selectedSigmas: [...args.deblurFold.selectedSigmas] };
  }
  if (args.styleTransferFold !== undefined && args.styleTransferFold !== null) {
    wire.styleTransferFold = {
      ...args.styleTransferFold,
      candidateLabels: [...args.styleTransferFold.candidateLabels],
      compositeScores: [...args.styleTransferFold.compositeScores],
      meanEffectiveGates: [...args.styleTransferFold.meanEffectiveGates],
    };
  }
  return wire;
}

/**
 * Handles the gnosis encode Mesh Carrier Assessment V1 Json workflow.
 */
export function encodeMeshCarrierAssessmentV1Json(
  args: EncodeMeshCarrierAssessmentV1Args,
): string {
  return JSON.stringify(toMeshCarrierAssessmentWire(args));
}

/**
 * Producer-side mirror of mesh-verifier's carrier Noise/Signal limit. Hosts can
 * inspect pressure locally, while mesh-verifier remains the alert writer.
 */
export function computeMeshCarrierNoiseSignalFromAssessments(
  assessments: readonly MeshCarrierAssessmentV1Wire[],
  spansProcessed = assessments.length,
): MeshCarrierNoiseSignalProjection {
  const denominator = Math.max(1, naturalCount(spansProcessed));
  let ready = 0;
  let notReady = 0;
  let recoverableWithWitness = 0;
  let partialObservation = 0;

  for (const assessment of assessments) {
    const horn = assessment.hornObservability;
    if (horn === undefined) {
      notReady += 1;
      continue;
    }
    if (horn.carrierReady) {
      ready += 1;
    } else {
      notReady += 1;
    }
    if (horn.recoverableWithWitness) {
      recoverableWithWitness += 1;
    }
    if (horn.partialObservation) {
      partialObservation += 1;
    }
  }

  const latest = assessments.at(-1);
  const readySignal = clamp01(ready / denominator);
  const recoverableSignal = clamp01(recoverableWithWitness / denominator);
  const signal = clamp01(readySignal + recoverableSignal);
  const notReadyNoise = clamp01(notReady / denominator);
  const partialObservationNoise = clamp01(partialObservation / denominator);
  const noise = clamp01(notReadyNoise + partialObservationNoise);
  const noiseSignalLimit = signal - noise;
  const latestHorn = latest?.hornObservability;
  const projection: MeshCarrierNoiseSignalProjection = {
    spansProcessed: denominator,
    certificatesObserved: assessments.length,
    certificateCoverage: clamp01(assessments.length / denominator),
    readySignal,
    recoverableSignal,
    signal,
    notReadyNoise,
    partialObservationNoise,
    noise,
    noiseSignalLimit,
    readinessMargin: latestHorn?.carrierReadiness ?? null,
    residuePressure:
      latestHorn === undefined
        ? null
        : pressure(latestHorn.unresolvedResidue),
    deficitPressure:
      latestHorn === undefined
        ? null
        : pressure(latestHorn.carrierReadinessDeficit),
    saturated: signal >= 1 || noise >= 1,
    interpretation:
      noiseSignalLimit < 0 ? 'noise-burdened' : 'signal-supported',
  };
  if (latestHorn !== undefined) {
    return {
      ...projection,
      theoremLineage: latestHorn.theoremLineage,
    };
  }
  return projection;
}

/** Maps an FRF deblur ensemble result into optional `deblurFold` telemetry. */
export function deblurEnsembleResultToMeshCarrierFold(
  result: DeblurEnsembleResult,
): MeshCarrierDeblurFoldWire {
  return {
    foldRounds: result.fold.rounds,
    foldSurvivors: result.fold.survivors,
    foldRejected: result.fold.rejected,
    foldTotalWeight: result.fold.totalWeight,
    candidateLabels: result.candidates.map((c) => c.label),
    selectedSigmas: result.candidates.map((c) => c.selectedSigma),
    watermark: result.invisibleWatermark,
  };
}

/**
 * Merges or replaces `deblurFold` on an existing serialized
 * `mesh_carrier_assessment/v1` JSON string (e.g. from `X-Mesh-Carrier-Assessment`).
 */
export function mergeDeblurFoldIntoMeshCarrierJson(
  existingJson: string,
  fold: MeshCarrierDeblurFoldWire,
): string {
  const parsed = JSON.parse(existingJson) as Record<string, unknown>;
  const v = parsed['certificateVersion'];
  if (v !== MESH_CARRIER_ASSESSMENT_CERTIFICATE_VERSION) {
    throw new Error(
      `mergeDeblurFoldIntoMeshCarrierJson: expected certificateVersion ${MESH_CARRIER_ASSESSMENT_CERTIFICATE_VERSION}, got ${String(v)}`,
    );
  }
  return JSON.stringify({ ...parsed, deblurFold: fold });
}

/**
 * Composes `deblurEnsembleResultToMeshCarrierFold` and
 * `mergeDeblurFoldIntoMeshCarrierJson` so a coordinator can attach
 * `forkRaceFoldDeblur` telemetry to an existing serialized assessment in one
 * step (same certificateVersion, no spectral re-encode).
 */
export function attachDeblurEnsembleToMeshCarrierAssessmentJson(
  meshAssessmentJson: string,
  ensemble: DeblurEnsembleResult,
): string {
  return mergeDeblurFoldIntoMeshCarrierJson(
    meshAssessmentJson,
    deblurEnsembleResultToMeshCarrierFold(ensemble),
  );
}

/** Maps an FRF style ensemble result into optional `styleTransferFold` telemetry. */
export function styleTransferEnsembleResultToMeshCarrierFold(
  result: StyleTransferEnsembleResult,
): MeshCarrierStyleTransferFoldWire {
  return {
    foldRounds: result.fold.rounds,
    foldSurvivors: result.fold.survivors,
    foldRejected: result.fold.rejected,
    foldTotalWeight: result.fold.totalWeight,
    candidateLabels: result.candidates.map((c) => c.label),
    compositeScores: result.candidates.map((c) => c.compositeFoldScore ?? c.styleDistance),
    meanEffectiveGates: result.candidates.map((c) => c.meshFoldMetrics?.mean_effective_gate ?? 1),
  };
}

/**
 * Handles the gnosis merge Style Transfer Fold Into Mesh Carrier Json workflow.
 */
export function mergeStyleTransferFoldIntoMeshCarrierJson(
  existingJson: string,
  fold: MeshCarrierStyleTransferFoldWire,
): string {
  const parsed = JSON.parse(existingJson) as Record<string, unknown>;
  const v = parsed['certificateVersion'];
  if (v !== MESH_CARRIER_ASSESSMENT_CERTIFICATE_VERSION) {
    throw new Error(
      `mergeStyleTransferFoldIntoMeshCarrierJson: expected certificateVersion ${MESH_CARRIER_ASSESSMENT_CERTIFICATE_VERSION}, got ${String(v)}`,
    );
  }
  return JSON.stringify({ ...parsed, styleTransferFold: fold });
}

/**
 * Handles the gnosis attach Style Transfer Ensemble To Mesh Carrier Assessment Json workflow.
 */
export function attachStyleTransferEnsembleToMeshCarrierAssessmentJson(
  meshAssessmentJson: string,
  ensemble: StyleTransferEnsembleResult,
): string {
  return mergeStyleTransferFoldIntoMeshCarrierJson(
    meshAssessmentJson,
    styleTransferEnsembleResultToMeshCarrierFold(ensemble),
  );
}

function utf8JsonToAsciiBase64(json: string): string {
  const bytes = new TextEncoder().encode(json);
  if (typeof Buffer !== 'undefined') {
    return Buffer.from(bytes).toString('base64');
  }
  let binary = '';
  for (let i = 0; i < bytes.length; i++) {
    binary += String.fromCharCode(bytes[i]!);
  }
  if (typeof globalThis.btoa !== 'function') {
    throw new Error('utf8JsonToAsciiBase64: no base64 encoder (btoa/Buffer)');
  }
  return globalThis.btoa(binary);
}

function naturalCount(value: number): number {
  if (!Number.isFinite(value)) return 0;
  return Math.max(0, Math.trunc(value));
}

function pressure(value: number): number {
  if (!Number.isFinite(value) || value <= 0) {
    return 0;
  }
  return value / (value + 1);
}

function clamp01(value: number): number {
  if (!Number.isFinite(value) || value <= 0) {
    return 0;
  }
  if (value >= 1) {
    return 1;
  }
  return value;
}

/**
 * Builds the worker request header value (base64 UTF-8 JSON) for
 * {@link MESH_HTTP_HEADER_CARRIER_DEBLUR_FOLD}. Coordinators that run
 * `forkRaceFoldDeblur` merge the fold onto mesh telemetry by forwarding this
 * header on `/forward`, `/embed`, `/lm-head`, etc.
 */
export function encodeDeblurFoldWireForMeshCarrierRequestHeader(
  fold: MeshCarrierDeblurFoldWire,
): string {
  const payload = JSON.stringify({
    foldRounds: fold.foldRounds,
    foldSurvivors: fold.foldSurvivors,
    foldRejected: fold.foldRejected,
    foldTotalWeight: fold.foldTotalWeight,
    candidateLabels: [...fold.candidateLabels],
    selectedSigmas: [...fold.selectedSigmas],
    watermark: fold.watermark,
  });
  return utf8JsonToAsciiBase64(payload);
}

/** Single-call header map for `fetch` to a layer worker after `forkRaceFoldDeblur`. */
export function meshCarrierRequestHeadersWithDeblurEnsemble(
  ensemble: DeblurEnsembleResult,
): Record<string, string> {
  const fold = deblurEnsembleResultToMeshCarrierFold(ensemble);
  return {
    [MESH_HTTP_HEADER_CARRIER_DEBLUR_FOLD]:
      encodeDeblurFoldWireForMeshCarrierRequestHeader(fold),
  };
}

/**
 * Handles the gnosis encode Style Transfer Fold Wire For Mesh Carrier Request Header workflow.
 */
export function encodeStyleTransferFoldWireForMeshCarrierRequestHeader(
  fold: MeshCarrierStyleTransferFoldWire,
): string {
  const payload = JSON.stringify({
    foldRounds: fold.foldRounds,
    foldSurvivors: fold.foldSurvivors,
    foldRejected: fold.foldRejected,
    foldTotalWeight: fold.foldTotalWeight,
    candidateLabels: [...fold.candidateLabels],
    compositeScores: [...fold.compositeScores],
    meanEffectiveGates: [...fold.meanEffectiveGates],
  });
  return utf8JsonToAsciiBase64(payload);
}

/**
 * Handles the gnosis mesh Carrier Request Headers With Style Transfer Ensemble workflow.
 */
export function meshCarrierRequestHeadersWithStyleTransferEnsemble(
  ensemble: StyleTransferEnsembleResult,
): Record<string, string> {
  const fold = styleTransferEnsembleResultToMeshCarrierFold(ensemble);
  return {
    [MESH_HTTP_HEADER_CARRIER_STYLE_FOLD]:
      encodeStyleTransferFoldWireForMeshCarrierRequestHeader(fold),
  };
}

/**
 * Handles the gnosis attention Closure Assessment To Wire workflow.
 */
export function attentionClosureAssessmentToWire(
  ac: MeshAttentionClosureAssessment,
): NonNullable<MeshCarrierAssessmentV1Wire['attentionClosure']> {
  return {
    meshAction: ac.action,
    decisionAction: ac.decision.action,
    decisionReason: ac.decision.reason,
    telemetryLabel: ac.workPolicy.telemetryLabel,
    theoremLineage: ac.workPolicy.theoremLineage,
    preserveHeads: ac.workPolicy.preserveHeads,
    observer: {
      heads: ac.observed.heads,
      coupling: ac.observed.coupling,
      observerResolution: ac.observed.observerResolution,
      kind: ac.observed.kind,
      unresolvedResidue: ac.observed.unresolvedResidue,
      traceSeparated: ac.observed.traceSeparated,
    },
  };
}
