// @a0n/gnosis/distributed-inference-host/src/mesh-federation.ts
import type { MeshModelConfig } from './mesh-model-registry';

export interface MeshFederationPeer {
  id: string;
  baseUrl: string;
  models: string[];
  priority: number;
  weight: number;
}

export interface MeshFederationLocalModel {
  id: string;
  modelName: string;
  aliases: string[];
  tokenizerFamily: MeshModelConfig['tokenizerFamily'];
  chatTemplate: MeshModelConfig['chatTemplate'];
}

export interface MeshFederationManifest {
  kind: 'gnosis-openai-mesh-manifest';
  version: 1;
  meshId: string;
  generatedAt: string;
  publicUrl?: string;
  loadedModel: string | null;
  childReady: boolean;
  localModels: MeshFederationLocalModel[];
  peers: MeshFederationPeer[];
}

type UnknownRecord = Record<string, unknown>;

function isRecord(value: unknown): value is UnknownRecord {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function requireString(value: unknown, name: string): string {
  if (typeof value !== 'string' || value.trim() === '') {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

function optionalNumber(value: unknown, name: string, fallback: number): number {
  if (value === undefined || value === null || value === '') return fallback;
  const parsed = typeof value === 'number' ? value : Number(value);
  if (!Number.isFinite(parsed) || parsed < 0) {
    throw new Error(`${name} must be a non-negative number`);
  }
  return parsed;
}

function assertPeerId(id: string): string {
  if (!/^[A-Za-z0-9._:-]{1,128}$/.test(id)) {
    throw new Error(`invalid mesh peer id "${id}"`);
  }
  return id;
}

function assertModelId(model: string): string {
  if (model === '*') return model;
  if (!/^[A-Za-z0-9._:/+-]{1,160}$/.test(model)) {
    throw new Error(`invalid mesh peer model "${model}"`);
  }
  return model;
}

function normalizeBaseUrl(raw: string): string {
  const parsed = new URL(raw);
  if (parsed.protocol !== 'http:' && parsed.protocol !== 'https:') {
    throw new Error(`mesh peer url must be http(s): ${raw}`);
  }
  parsed.hash = '';
  parsed.search = '';
  return parsed.toString().replace(/\/$/, '');
}

function derivePeerId(url: string): string {
  const host = new URL(url).hostname.replace(/[^A-Za-z0-9._:-]/g, '-');
  return assertPeerId(host.slice(0, 128) || 'mesh-peer');
}

function splitList(raw: string, separator: string): string[] {
  return raw
    .split(separator)
    .map((piece) => piece.trim())
    .filter((piece) => piece.length > 0);
}

function parseJsonPeer(value: unknown, index: number): MeshFederationPeer {
  if (!isRecord(value)) throw new Error(`MESH_FEDERATION_PEERS[${index}] must be an object`);
  const baseUrl = normalizeBaseUrl(requireString(value['baseUrl'] ?? value['url'], 'baseUrl'));
  const id = assertPeerId(
    typeof value['id'] === 'string' && value['id'].trim() !== ''
      ? value['id'].trim()
      : derivePeerId(baseUrl),
  );
  const rawModels = value['models'];
  const models =
    Array.isArray(rawModels) && rawModels.length > 0
      ? rawModels.map((model, modelIndex) =>
          assertModelId(requireString(model, `models[${modelIndex}]`)),
        )
      : ['*'];
  return {
    id,
    baseUrl,
    models,
    priority: optionalNumber(value['priority'], 'priority', 100),
    weight: optionalNumber(value['weight'], 'weight', 1),
  };
}

function parseCompactPeer(raw: string): MeshFederationPeer {
  const [left, modelPart] = raw.split('#', 2);
  const head = requireString(left, 'mesh peer');
  const atIndex = head.indexOf('@');
  const explicitId = atIndex > 0 ? head.slice(0, atIndex).trim() : '';
  const rawUrl = atIndex > 0 ? head.slice(atIndex + 1).trim() : head;
  const baseUrl = normalizeBaseUrl(rawUrl);
  const models =
    modelPart && modelPart.trim() !== ''
      ? splitList(modelPart, '|').map(assertModelId)
      : ['*'];
  return {
    id: explicitId ? assertPeerId(explicitId) : derivePeerId(baseUrl),
    baseUrl,
    models,
    priority: 100,
    weight: 1,
  };
}

/**
 * Parses the Mesh Federation Peers.
 */
export function parseMeshFederationPeers(raw: string | undefined): MeshFederationPeer[] {
  if (raw == null || raw.trim() === '') return [];
  const trimmed = raw.trim();
  if (trimmed.startsWith('[')) {
    const parsed = JSON.parse(trimmed) as unknown;
    if (!Array.isArray(parsed)) throw new Error('MESH_FEDERATION_PEERS JSON must be an array');
    return parsed.map(parseJsonPeer);
  }
  return splitList(trimmed, ',').map(parseCompactPeer);
}

/**
 * Handles the gnosis local Model Capabilities workflow.
 */
export function localModelCapabilities(
  models: readonly MeshModelConfig[],
): MeshFederationLocalModel[] {
  return models.map((model) => ({
    id: model.id,
    modelName: model.modelName,
    aliases: model.aliases.slice(),
    tokenizerFamily: model.tokenizerFamily,
    chatTemplate: model.chatTemplate,
  }));
}

/**
 * Builds the Mesh Federation Manifest.
 */
export function buildMeshFederationManifest(input: {
  meshId: string;
  publicUrl?: string;
  loadedModel: string | null;
  childReady: boolean;
  models: readonly MeshModelConfig[];
  peers: readonly MeshFederationPeer[];
  generatedAt?: string;
}): MeshFederationManifest {
  const manifest: MeshFederationManifest = {
    kind: 'gnosis-openai-mesh-manifest',
    version: 1,
    meshId: assertPeerId(input.meshId),
    generatedAt: input.generatedAt ?? new Date().toISOString(),
    loadedModel: input.loadedModel,
    childReady: input.childReady,
    localModels: localModelCapabilities(input.models),
    peers: input.peers.map((peer) => ({
      id: peer.id,
      baseUrl: peer.baseUrl,
      models: peer.models.slice(),
      priority: peer.priority,
      weight: peer.weight,
    })),
  };
  if (input.publicUrl !== undefined && input.publicUrl.trim() !== '') {
    manifest.publicUrl = normalizeBaseUrl(input.publicUrl);
  }
  return manifest;
}

/**
 * Resolves the Federated Peer For Model.
 */
export function resolveFederatedPeerForModel(
  requestedModel: string,
  peers: readonly MeshFederationPeer[],
): MeshFederationPeer | null {
  const requested = requestedModel.trim().toLowerCase();
  if (requested === '') return null;
  const candidates = peers
    .filter((peer) =>
      peer.models.some((model) => model === '*' || model.toLowerCase() === requested),
    )
    .sort((a, b) => a.priority - b.priority || b.weight - a.weight || a.id.localeCompare(b.id));
  return candidates[0] ?? null;
}
