// @a0n/gnosis/distributed-inference-host/src/native-tool-formats.ts
/**
 * Per-model NATIVE tool-calling wire formats — Qwen2.5/Qwen3, Mistral V3, and
 * (experimental, gated) Gemma4 — replacing the generic prompt-injected
 * fenced-block convention (`agentic-chat.ts`'s `formatToolsForPrompt`/
 * `parseToolCalls`) with each model's own trained tool-calling tokens, for the
 * chat-template families where that format has been independently verified.
 *
 * Every function here is pure (string/array in, string/array out) so the
 * whole module is unit-testable without a network call or a running model.
 */

import type { AgenticToolCall, AgenticToolDefinition } from './agentic-chat';

export type NativeToolFamily = 'qwen' | 'mistral-v3' | 'gemma4';

/**
 * `chatTemplate` -> native tool-calling family, or `null` to keep today's
 * generic fenced-block convention. Qwen and Mistral are default-eligible;
 * Gemma4 requires an explicit opt-in (`opts.gemma4Enabled`) since its only
 * currently-registered models (`gemma4-12b-it`/`gemma4-31b-it`) load a
 * gemma-3 tokenizer that lacks the required special tokens — enabling this
 * without a real gemma4-vocab tokenizer asset would produce byte-fallback
 * garbage, the same failure mode already documented in
 * `mesh-model-registry.ts` for plain chat.
 */
export function resolveNativeToolFamily(
  chatTemplate: string,
  opts?: { gemma4Enabled?: boolean },
): NativeToolFamily | null {
  switch (chatTemplate) {
    case 'qwen-chatml':
      return 'qwen';
    case 'mistral-v3-instruct':
      return 'mistral-v3';
    case 'gemma4':
      return opts?.gemma4Enabled ? 'gemma4' : null;
    default:
      return null;
  }
}

export interface ParsedNativeToolCalls {
  text: string;
  toolCalls: AgenticToolCall[] | null;
}

// ---------------------------------------------------------------------------
// Shared helpers
// ---------------------------------------------------------------------------

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function createCallId(): string {
  if (typeof crypto !== 'undefined' && typeof crypto.randomUUID === 'function') {
    return `call_${crypto.randomUUID().slice(0, 12)}`;
  }
  return `call_${Date.now().toString(36)}_${Math.random().toString(36).slice(2, 8)}`;
}

/** `AgenticToolCall.function.arguments` is always a JSON-encoded string
 * (OpenAI convention) — parse it back to a real value for re-serialization
 * into a native format that expects an object, not a string-of-a-string. */
function safeParseArguments(argumentsJson: string): unknown {
  try {
    return JSON.parse(argumentsJson);
  } catch {
    return {};
  }
}

function safeParseJsonOrString(value: string): unknown {
  try {
    return JSON.parse(value);
  } catch {
    return value;
  }
}

function parseJsonToolCallObject(
  raw: string,
): { name: string; arguments: unknown } | null {
  try {
    const value: unknown = JSON.parse(raw);
    if (!isPlainObject(value)) return null;
    const name = value.name;
    if (typeof name !== 'string' || name.length === 0) return null;
    return { name, arguments: value.arguments ?? {} };
  } catch {
    return null;
  }
}

// ---------------------------------------------------------------------------
// Qwen2.5 / Qwen3 — verified against the real `tokenizer_config.json`
// chat_template Jinja source for Qwen/Qwen2.5-7B-Instruct and Qwen/Qwen3-8B
// (identical tool-calling tags between the two). chatTemplate: 'qwen-chatml'.
// ---------------------------------------------------------------------------

/**
 * Formats the Qwen Tools System Block.
 */
export function formatQwenToolsSystemBlock(tools: AgenticToolDefinition[]): string {
  if (tools.length === 0) return '';
  const toolLines = tools
    .map((tool) => JSON.stringify({ type: 'function', function: tool.function }))
    .join('\n');
  return `# Tools

You may call one or more functions to assist with the user query.

You are provided with function signatures within <tools></tools> XML tags:
<tools>
${toolLines}
</tools>

For each function call, return a json object with function name and arguments within <tool_call></tool_call> XML tags:
<tool_call>
{"name": <function-name>, "arguments": <args-json-object>}
</tool_call>`;
}

const QWEN_TOOL_CALL_RE = /<tool_call>([\s\S]*?)<\/tool_call>/gu;

/**
 * Parses the Qwen Native Tool Calls.
 */
export function parseQwenNativeToolCalls(
  output: string,
  validToolNames?: Set<string>,
): ParsedNativeToolCalls | null {
  const matches = [...output.matchAll(QWEN_TOOL_CALL_RE)];
  if (matches.length === 0) return null; // no native markers at all -> caller falls back to the generic parser

  const toolCalls: AgenticToolCall[] = [];
  let text = output;
  for (const match of matches) {
    const parsed = parseJsonToolCallObject((match[1] ?? '').trim());
    if (parsed && (!validToolNames || validToolNames.has(parsed.name))) {
      toolCalls.push({
        id: createCallId(),
        type: 'function',
        function: { name: parsed.name, arguments: JSON.stringify(parsed.arguments) },
      });
    }
    text = text.replace(match[0], '').trim();
  }
  return { text: text.trim(), toolCalls: toolCalls.length > 0 ? toolCalls : null };
}

/** Renders a PAST assistant `tool_calls` message back into Qwen's native
 * shape, for re-rendering conversation history on a later turn. */
export function formatQwenAssistantToolCallTurn(calls: AgenticToolCall[]): string {
  return calls
    .map(
      (call) =>
        `<tool_call>\n${JSON.stringify({
          name: call.function.name,
          arguments: safeParseArguments(call.function.arguments),
        })}\n</tool_call>`,
    )
    .join('\n');
}

/**
 * Formats the Qwen Tool Result Turn.
 */
export function formatQwenToolResultTurn(resultContent: string): string {
  return `<tool_response>\n${resultContent}\n</tool_response>`;
}

// ---------------------------------------------------------------------------
// Mistral V3 (Mistral-7B-Instruct-v0.3, Codestral-22B) — verified from
// Mistral's tokenization/tool_calling cookbook + the official HF model card
// confirming v0.3 uses the V3 (with `id`/`call_id`) convention, not the
// older ID-less V2 one. chatTemplate: 'mistral-v3-instruct'.
// ---------------------------------------------------------------------------

/** `[AVAILABLE_TOOLS] [...][/AVAILABLE_TOOLS]` — a TOP-LEVEL prompt segment
 * that sits BEFORE the first `[INST]` turn, not inside any message's
 * content. Consumed directly by `applyMistralV3InstructTemplate`. */
export function buildMistralAvailableToolsSegment(tools: AgenticToolDefinition[]): string {
  if (tools.length === 0) return '';
  const toolsJson = JSON.stringify(
    tools.map((tool) => ({ type: 'function', function: tool.function })),
  );
  return `[AVAILABLE_TOOLS] ${toolsJson}[/AVAILABLE_TOOLS]`;
}

const MISTRAL_CALL_ID_ALPHABET =
  'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';

/** 9-character alphanumeric id, per Mistral's V3 tool-calling convention
 * ("They should be exactly 9 alphanumeric characters" — HF model card). */
export function generateMistralToolCallId(random: () => number = Math.random): string {
  let id = '';
  for (let i = 0; i < 9; i += 1) {
    id += MISTRAL_CALL_ID_ALPHABET[Math.floor(random() * MISTRAL_CALL_ID_ALPHABET.length)];
  }
  return id;
}

const MISTRAL_TOOL_CALLS_RE = /\[TOOL_CALLS\]\s*(\[[\s\S]*?\])/u;

/**
 * Parses the Mistral Native Tool Calls.
 */
export function parseMistralNativeToolCalls(
  output: string,
  validToolNames?: Set<string>,
): ParsedNativeToolCalls | null {
  const match = MISTRAL_TOOL_CALLS_RE.exec(output);
  if (!match) return null; // no native marker at all -> caller falls back to the generic parser

  const text = output.replace(match[0], '').trim();
  let parsedArray: unknown;
  try {
    parsedArray = JSON.parse(match[1]!);
  } catch {
    // Native marker present but unparsable (truncated/malformed) — a real
    // marker was found, so don't fall through to the generic parser (which
    // would fail identically on this same malformed text).
    return { text, toolCalls: null };
  }
  if (!Array.isArray(parsedArray)) {
    return { text, toolCalls: null };
  }

  const toolCalls: AgenticToolCall[] = [];
  for (const entry of parsedArray) {
    if (!isPlainObject(entry)) continue;
    const name = entry.name;
    if (typeof name !== 'string' || name.length === 0) continue;
    if (validToolNames && !validToolNames.has(name)) continue;
    // Preserve the MODEL'S OWN id verbatim so the result round-trip
    // (formatMistralToolResultSegment) can reference the matching call_id
    // with zero extra tracking state.
    const id =
      typeof entry.id === 'string' && entry.id.length > 0
        ? entry.id
        : generateMistralToolCallId();
    toolCalls.push({
      id,
      type: 'function',
      function: { name, arguments: JSON.stringify(entry.arguments ?? {}) },
    });
  }
  return { text, toolCalls: toolCalls.length > 0 ? toolCalls : null };
}

/** Renders a PAST assistant `tool_calls` message back into Mistral's native
 * `[TOOL_CALLS] [...]` shape (used both for the live turn and for re-rendering
 * history), preserving each call's own `id`. */
export function formatMistralToolCallsSegment(calls: AgenticToolCall[]): string {
  const array = calls.map((call) => ({
    name: call.function.name,
    arguments: safeParseArguments(call.function.arguments),
    id: call.id,
  }));
  return `[TOOL_CALLS] ${JSON.stringify(array)}`;
}

/** `[TOOL_RESULTS] {"content": ..., "call_id": ...}[/TOOL_RESULTS]` — a
 * SINGLE object (not a list, unlike the older V2 convention), field name is
 * `call_id` (matching the `id` from the originating tool call). */
export function formatMistralToolResultSegment(resultContent: string, callId: string): string {
  return `[TOOL_RESULTS] ${JSON.stringify({
    content: resultContent,
    call_id: callId,
  })}[/TOOL_RESULTS]`;
}

/** Minimal message shape needed to render a full Mistral V3 prompt —
 * intentionally NOT importing `openai-server.ts`'s `Message` type (this
 * module must stay a leaf with no dependency on that entrypoint file, which
 * cannot be safely imported in tests — see `renderMistralV3InstructPrompt`'s
 * doc comment). */
export interface MistralRenderableMessage {
  role: string;
  content: string;
  tool_calls?: AgenticToolCall[];
  tool_call_id?: string;
}

/**
 * Full Mistral V3 prompt renderer, including the tool round-trip. Lives here
 * (not in `openai-server.ts`) specifically so it's exact-string unit
 * testable: `openai-server.ts` calls `main()` unconditionally at module
 * scope with no entrypoint guard, so it can never be safely `import`-ed by a
 * test — this is the one piece of that file's chat-template logic complex
 * enough to need real (not source-string-presence) tests, so the logic
 * itself lives in this already-safe, already-tested module instead.
 *
 * A sequential walk (not the original two-pass user/assistant turn-pairing)
 * — required because Mistral's real wire format inserts `[TOOL_CALLS]...`/
 * `[TOOL_RESULTS]...[/TOOL_RESULTS]` INSIDE an already-open exchange, which
 * the turn-pairing model has no slot for. Verified byte-for-byte against
 * Mistral's own cookbook example for the no-tool-calls case (see
 * `native-tool-formats.test.ts`).
 */
export function renderMistralV3InstructPrompt(
  messages: MistralRenderableMessage[],
  tools?: AgenticToolDefinition[],
): string {
  let systemPrompt = '';
  for (const msg of messages) {
    if (msg.role === 'system') systemPrompt = msg.content;
  }

  let result = tools && tools.length > 0 ? buildMistralAvailableToolsSegment(tools) : '';
  let firstUserSeen = false;

  for (const msg of messages) {
    if (msg.role === 'system') continue;

    if (msg.role === 'user') {
      let content = msg.content;
      if (!firstUserSeen && systemPrompt) {
        content = `${systemPrompt}\n\n${content}`;
      }
      firstUserSeen = true;
      result += `[INST] ${content} [/INST]`;
      continue;
    }

    if (msg.role === 'assistant' && msg.tool_calls && msg.tool_calls.length > 0) {
      // The model's tool-call output closes with </s> immediately (its own
      // generation ends there) — [TOOL_RESULTS] and the eventual final
      // answer are NOT wrapped in a new [INST] block.
      result += `${formatMistralToolCallsSegment(msg.tool_calls)}</s>`;
      continue;
    }

    if (msg.role === 'tool') {
      result += formatMistralToolResultSegment(msg.content, msg.tool_call_id ?? '');
      continue;
    }

    if (msg.role === 'assistant') {
      result += ` ${msg.content}</s>`;
      continue;
    }
  }

  return result;
}

// ---------------------------------------------------------------------------
// Gemma4 — EXPERIMENTAL, gated behind `GEMMA4_NATIVE_TOOLS_EXPERIMENTAL`.
// Verified from the live Gemma4 function-calling doc, but NOT wired to the
// currently-registered `gemma4-12b-it`/`gemma4-31b-it` ids (their loaded
// gemma-3 tokenizer lacks these special tokens — see resolveNativeToolFamily
// and mesh-model-registry.ts's documented fix). No live smoke test exists yet
// for this family; treat this section as implemented-but-unverified-live.
//
// Gemma4's own output format is a custom unquoted-key mini-DSL, NOT JSON
// (string literals wrapped in `<|"|>...<|"|>` instead of `"..."`). The
// parser below converts the DSL to valid JSON via string substitution before
// calling JSON.parse — this does NOT escape a literal `"` inside a wrapped
// string value, a known limitation acceptable given this family has no live
// traffic yet.
// ---------------------------------------------------------------------------

function gemma4Serialize(value: unknown): string {
  if (value === null || value === undefined) return 'null';
  if (typeof value === 'string') return `<|"|>${value}<|"|>`;
  if (typeof value === 'number' || typeof value === 'boolean') return String(value);
  if (Array.isArray(value)) return `[${value.map(gemma4Serialize).join(',')}]`;
  if (isPlainObject(value)) {
    return `{${Object.entries(value)
      .map(([key, val]) => `${key}:${gemma4Serialize(val)}`)
      .join(',')}}`;
  }
  return `<|"|>${String(value)}<|"|>`;
}

/**
 * Formats the Gemma4 Tool Declaration Block.
 */
export function formatGemma4ToolDeclarationBlock(tools: AgenticToolDefinition[]): string {
  return tools
    .map((tool) => {
      const body = gemma4Serialize({
        description: tool.function.description ?? '',
        parameters: tool.function.parameters ?? { type: 'object', properties: {} },
      });
      return `<|tool>declaration:${tool.function.name}${body}<tool|>`;
    })
    .join('\n');
}

function gemma4DslToJson(dsl: string): string {
  const withQuotedStrings = dsl.replace(/<\|"\|>/gu, '"');
  return withQuotedStrings.replace(/([{,]\s*)([A-Za-z_][A-Za-z0-9_]*)\s*:/gu, '$1"$2":');
}

function parseGemma4Body(body: string): Record<string, unknown> | null {
  try {
    const value: unknown = JSON.parse(gemma4DslToJson(body));
    return isPlainObject(value) ? value : null;
  } catch {
    return null;
  }
}

/** Balanced-brace scan starting at a `{` — a non-greedy regex would stop at
 * the first nested `}` and truncate. Returns the full `{...}` slice
 * (including braces), or null if unbalanced/truncated. */
function extractBalancedBraces(input: string, openIndex: number): string | null {
  if (input[openIndex] !== '{') return null;
  let depth = 0;
  for (let i = openIndex; i < input.length; i += 1) {
    if (input[i] === '{') depth += 1;
    else if (input[i] === '}') {
      depth -= 1;
      if (depth === 0) return input.slice(openIndex, i + 1);
    }
  }
  return null;
}

const GEMMA4_TOOL_CALL_OPEN_RE = /<\|tool_call>call:([A-Za-z_][A-Za-z0-9_./-]*)\{/gu;
const GEMMA4_TOOL_CALL_CLOSE = '<tool_call|>';

/**
 * Parses the Gemma4 Native Tool Calls.
 */
export function parseGemma4NativeToolCalls(
  output: string,
  validToolNames?: Set<string>,
): ParsedNativeToolCalls | null {
  const opens = [...output.matchAll(GEMMA4_TOOL_CALL_OPEN_RE)];
  if (opens.length === 0) return null; // no native markers at all -> caller falls back to the generic parser

  const toolCalls: AgenticToolCall[] = [];
  let text = output;
  for (const open of opens) {
    const name = open[1]!;
    const braceStart = open.index! + open[0].length - 1;
    const body = extractBalancedBraces(output, braceStart);
    if (!body) continue; // truncated/unbalanced — skip this one, keep scanning others
    const afterBrace = output.slice(
      braceStart + body.length,
      braceStart + body.length + GEMMA4_TOOL_CALL_CLOSE.length,
    );
    const fullMatchEnd =
      braceStart + body.length + (afterBrace === GEMMA4_TOOL_CALL_CLOSE ? GEMMA4_TOOL_CALL_CLOSE.length : 0);
    const fullMatch = output.slice(open.index!, fullMatchEnd);
    const parsedArgs = parseGemma4Body(body);
    if (parsedArgs !== null && (!validToolNames || validToolNames.has(name))) {
      toolCalls.push({
        id: createCallId(),
        type: 'function',
        function: { name, arguments: JSON.stringify(parsedArgs) },
      });
    }
    text = text.replace(fullMatch, '').trim();
  }
  return { text: text.trim(), toolCalls: toolCalls.length > 0 ? toolCalls : null };
}

/**
 * Formats the Gemma4 Tool Call Turn.
 */
export function formatGemma4ToolCallTurn(calls: AgenticToolCall[]): string {
  return calls
    .map(
      (call) =>
        `<|tool_call>call:${call.function.name}${gemma4Serialize(
          safeParseArguments(call.function.arguments),
        )}<tool_call|>`,
    )
    .join('\n');
}

/**
 * Formats the Gemma4 Tool Result Turn.
 */
export function formatGemma4ToolResultTurn(resultContent: string, toolName: string): string {
  const parsedResult = safeParseJsonOrString(resultContent);
  const body = isPlainObject(parsedResult)
    ? gemma4Serialize(parsedResult)
    : gemma4Serialize({ result: parsedResult });
  return `<|tool_response>response:${toolName}${body}<tool_response|>`;
}

// ---------------------------------------------------------------------------
// Dispatchers — the only entry points agentic-chat.ts / openai-server.ts need
// ---------------------------------------------------------------------------

/**
 * Parses the Native Tool Calls.
 */
export function parseNativeToolCalls(
  family: NativeToolFamily,
  output: string,
  validToolNames?: Set<string>,
): ParsedNativeToolCalls | null {
  switch (family) {
    case 'qwen':
      return parseQwenNativeToolCalls(output, validToolNames);
    case 'mistral-v3':
      return parseMistralNativeToolCalls(output, validToolNames);
    case 'gemma4':
      return parseGemma4NativeToolCalls(output, validToolNames);
  }
}

/** `systemAddition` is appended to the system message content (Qwen, Gemma4);
 * `preamble` is a top-level segment the Mistral renderer must place before
 * the first `[INST]` turn (never both set for a given family). */
export function formatNativeToolsDeclaration(
  family: NativeToolFamily,
  tools: AgenticToolDefinition[],
): { systemAddition?: string; preamble?: string } {
  switch (family) {
    case 'qwen':
      return { systemAddition: formatQwenToolsSystemBlock(tools) };
    case 'mistral-v3':
      return { preamble: buildMistralAvailableToolsSegment(tools) };
    case 'gemma4':
      return { systemAddition: formatGemma4ToolDeclarationBlock(tools) };
  }
}
