// @a0n/gnosis/distributed-inference-host/src/ngram-draft.ts
/**
 * ngram-draft.ts — prompt-lookup / n-gram speculative-decode DRAFT (no model).
 *
 * Standard prompt-lookup decoding (Saxena 2023): instead of running a small
 * draft MODEL, we look up the running token sequence itself. To draft the
 * tokens that should come after the current position, find the longest suffix
 * of `seq` (length n, trying n = nMax .. 1) that occurred EARLIER in `seq`, and
 * propose the up-to-K tokens that FOLLOWED that earlier occurrence.
 *
 * This is cheap (string-matching over the token buffer), needs no model and no
 * training, and is highly effective on repetitive / structured output (code,
 * JSON, repeated phrases) — exactly where the verify-batch acceptance rate is
 * high. On novel prose it usually finds no match → the loop falls back to one
 * normal decode step (baseline).
 *
 * Pure function. The verify primitive (Rust `/verify-batch` → `inner_verify_batch`)
 * is what makes the proposed tokens LOSSLESS for greedy decoding: a drafted
 * token is emitted only if it equals the target model's argmax at that slot.
 */

export interface NgramDraftOptions {
  /** Max draft length K (tokens to propose). Default 4. */
  readonly k?: number;
  /** Max n-gram suffix length to match on. Default 3. */
  readonly nMax?: number;
  /**
   * Minimum suffix length to consider (default 1). Raising this trades recall
   * for precision — only longer, more specific repeats draft.
   */
  readonly nMin?: number;
}

/** Result of {@link ngramDraftWithMatch}: the draft + how strong the match was. */
export interface NgramDraftMatch {
  /** Up-to-K proposed continuation tokens. Empty when no suffix recurs. */
  readonly draft: number[];
  /**
   * Length n of the recurring suffix that produced `draft` (the n-gram order of
   * the match). 0 when `draft` is empty. A LONGER prior match is a stronger
   * signal: the continuation has a higher continuation probability and so a
   * higher verify acceptance rate. The cost-gated caller uses this to suppress
   * weak/short matches (which waste a K-token verify on near-zero acceptance).
   */
  readonly matchLen: number;
}

/**
 * Propose up to `k` draft tokens for the position after the end of `seq`, using
 * prompt-lookup / n-gram matching, AND report the length of the suffix match
 * that produced them. Returns `{ draft: [], matchLen: 0 }` when no suffix of
 * `seq` (length nMin..nMax) recurs earlier with at least one follower — the
 * caller MUST then fall back to a single normal decode step.
 *
 * Matching strategy: longest suffix first (n = nMax .. nMin); within a fixed n,
 * the LATEST earlier occurrence (rightmost, closest to the cursor) is used,
 * since recent context is the most predictive. The proposed span is the up-to-K
 * tokens immediately following that occurrence (truncated at the end of `seq`).
 */
export function ngramDraftWithMatch(
  seq: ArrayLike<number>,
  options: NgramDraftOptions = {},
): NgramDraftMatch {
  const k = options.k ?? 4;
  const nMax = options.nMax ?? 3;
  const nMin = options.nMin ?? 1;
  const len = seq.length;
  if (len === 0 || k <= 0) return { draft: [], matchLen: 0 };
  if (!Number.isInteger(k) || !Number.isInteger(nMax) || !Number.isInteger(nMin)) {
    throw new Error('ngramDraft: k, nMax, nMin must be integers');
  }
  if (nMin < 1 || nMax < nMin) {
    throw new Error(
      `ngramDraft: require 1 <= nMin <= nMax (got nMin=${nMin}, nMax=${nMax})`,
    );
  }

  const hi = Math.min(nMax, len);
  for (let n = hi; n >= nMin; n--) {
    // The earlier occurrence must leave room for >= 1 follower, so its start
    // index is in [0, len - n - 1]. Search rightmost-first (latest occurrence).
    if (len < n + 1) continue;
    const suffixStart = len - n;
    for (let start = len - n - 1; start >= 0; start--) {
      let matched = true;
      for (let j = 0; j < n; j++) {
        if (seq[start + j] !== seq[suffixStart + j]) {
          matched = false;
          break;
        }
      }
      if (!matched) continue;
      const followStart = start + n;
      const take = Math.min(k, len - followStart);
      if (take <= 0) continue; // suffix recurred but with no follower; keep searching
      const out: number[] = new Array(take);
      for (let i = 0; i < take; i++) out[i] = seq[followStart + i]!;
      return { draft: out, matchLen: n };
    }
  }
  return { draft: [], matchLen: 0 };
}

/**
 * Propose up to `k` draft tokens for the position after the end of `seq`, using
 * prompt-lookup / n-gram matching. Returns an empty array when no suffix of
 * `seq` (length nMin..nMax) recurs earlier with at least one follower — the
 * caller MUST then fall back to a single normal decode step.
 *
 * Thin wrapper over {@link ngramDraftWithMatch} that discards the match length.
 */
export function ngramDraft(
  seq: ArrayLike<number>,
  options: NgramDraftOptions = {},
): number[] {
  return ngramDraftWithMatch(seq, options).draft;
}
