// @a0n/gnosis/distributed-inference-host/src/occupancy-count.ts
/**
 * # Occupancy Count — how many people are in the room, read from room-tone modulation
 *
 * Mirrors `Gnosis.Mesh.OccupancyCount` from `open-source/gnosis-math`, the Rustic
 * Church carrier for *room occupancy estimation*: from a window of 1-D modulation
 * observations (each a scalar feature — a quantized breathing-rate band,
 * micro-motion cadence, or slow envelope perturbation above the steady ambient
 * carrier), estimate HOW MANY people are present. Every occupant perturbs the
 * steady room-tone carrier with a distinct slow modulation signature; once the
 * EMA baseline is subtracted, each present body leaves a cluster of modulation
 * observations clustered around its own signature. The estimate is not a
 * free-floating number; it is the *length of an admitted clustering*. The Skeptic
 * admits an occupancy count only when the modulation observations partition into
 * that many groups that are
 *
 *  - **internally tight** — every member of a group sits within `intraTol` of the
 *    group's representative (its head), so the group is one person's modulation
 *    spread and not two people's signatures smeared together;
 *  - **mutually separated** — any two distinct occupants' representatives differ
 *    by more than `sepThreshold`, so the split is not a phantom (one person's
 *    noise carved into two "people"); and
 *  - **supported** — every group has at least `minMembers` members, so it is a
 *    real occupant rather than a single stray sub-floor perturbation that never
 *    cleared the modulation support floor (an empty patch of carrier, not a body).
 *
 * ## What is (and is not) carried
 *
 * Two failure modes are refuted, both about the *count*:
 *
 *  1. **Over-merge (under-count):** two well-separated occupants called one
 *     person. The merged group's internal spread blows past `intraTol`, so the
 *     1-cluster proposal fails the tightness gate.
 *  2. **Over-split (over-count):** one person's natural modulation jitter carved
 *     into several "people". Two representatives sit within `sepThreshold`, so the
 *     proposal fails the separation gate.
 *
 * The support gate sharpens the occupancy reading to a person-vs-noise modulation
 * floor: a lone sub-floor cluster (a single perturbation that never reached the
 * floor) is carrier noise, not a body, and is rejected on support.
 *
 * This module is the runtime promotion check for an occupancy-count observation.
 * It carries ONLY the finite contract the Lean file proves: the saturating
 * feature metric (`adist`), the three gates (`tightness` / `wellSeparated` /
 * `supported`) and their aggregate admission (`admitOccupancy`). Real EMA
 * baseline tracking, micro-Doppler / radar modulation extraction, breathing-band
 * estimation, presence classification, and crowd-density model-order selection
 * stay runtime/empirical obligations OUTSIDE this certificate.
 *
 * The promotion check folds the contract into the same epistemic-outcome
 * vocabulary the sibling mesh carriers already speak
 * (`claim_admitted` / `observation_only` / `claim_withheld`), so an
 * occupancy-count claim cannot be displayed until it has cleared the Lean-side
 * gates. The empty room (no modulation above baseline) is the degenerate
 * observation-only case: an honest count of 0, no one present.
 *
 * Representative convention: a group's representative is its head. The empty
 * group has no representative (`0`) and is treated as unsupported.
 *
 * ## Lean source-of-truth
 *
 * @see open-source/gnosis-math/Gnosis/Mesh/OccupancyCount.lean
 * @see open-source/gnosis-math/Gnosis/Mesh/AcousticSourceCounting.lean (clustering sibling)
 */

/* ------------------------------------------------------------------ */
/*  Modulation feature distance, observations, clusters               */
/* ------------------------------------------------------------------ */

/** An observation is a 1-D modulation feature above the room-tone baseline
 *  (a quantized breathing-rate band or micro-motion cadence scalar). */
export type Observation = number;

/** A window of modulation observations over a frame. */
export type Window = readonly Observation[];

/** A cluster (proposed occupant group) is a list of observations; its
 *  representative is its head. */
export type Cluster = readonly Observation[];

/** A proposed occupancy clustering: the list of occupant groups. The admitted
 *  occupancy count is its length. */
export type Clustering = readonly Cluster[];

/** Absolute difference on `Nat` via saturating subtraction (`|a - b|`). The 1-D
 *  modulation-feature residual metric, shared with the mesh corridor carriers.
 *  Mirrors `OccupancyCount.adist`. */
export function adist(a: number, b: number): number {
  return Math.max(0, a - b) + Math.max(0, b - a);
}

/** The representative of a cluster: its head, or `0` for the empty cluster (which
 *  is independently rejected by the support gate). Mirrors `OccupancyCount.repOf`. */
export function repOf(group: Cluster): number {
  return group.length === 0 ? 0 : group[0];
}

/* ------------------------------------------------------------------ */
/*  Gate 1: intra-cluster tightness                                   */
/* ------------------------------------------------------------------ */

/**
 * A cluster is tight when every member is within `intraTol` of its representative
 * (its head). The over-merge under-count failure is exactly a tightness failure.
 * Mirrors `OccupancyCount.clusterTightB`.
 */
export function tightness(intraTol: number, group: Cluster): boolean {
  const rep = repOf(group);
  for (const a of group) {
    if (adist(a, rep) > intraTol) {
      return false;
    }
  }
  return true;
}

/** Every cluster in a clustering is tight. Mirrors `OccupancyCount.allTightB`. */
export function allTight(intraTol: number, clusters: Clustering): boolean {
  return clusters.every((g) => tightness(intraTol, g));
}

/* ------------------------------------------------------------------ */
/*  Gate 2: inter-cluster separation                                  */
/* ------------------------------------------------------------------ */

/** The list of representatives of a clustering. Mirrors `OccupancyCount.repsOf`. */
export function repsOf(clusters: Clustering): number[] {
  return clusters.map((g) => repOf(g));
}

/**
 * The clustering is well-separated when its representatives are pairwise more
 * than `sepThreshold` apart. Two clusters within `sepThreshold` mark an
 * over-split (one person double-counted) and fail this gate.
 * Mirrors `OccupancyCount.wellSeparatedB`.
 */
export function wellSeparated(
  sepThreshold: number,
  clusters: Clustering,
): boolean {
  const reps = repsOf(clusters);
  for (let i = 0; i < reps.length; i++) {
    for (let j = i + 1; j < reps.length; j++) {
      if (!(adist(reps[i], reps[j]) > sepThreshold)) {
        return false;
      }
    }
  }
  return true;
}

/* ------------------------------------------------------------------ */
/*  Gate 3: modulation support floor                                  */
/* ------------------------------------------------------------------ */

/** A cluster has support when it has at least `minMembers` modulation
 *  observations: a real occupant leaves a sustained modulation cluster, whereas a
 *  single sub-floor perturbation is carrier noise, not a person.
 *  Mirrors `OccupancyCount.supportedB`. */
export function supportedCluster(minMembers: number, group: Cluster): boolean {
  return group.length >= minMembers;
}

/** Every cluster in a clustering has support. Mirrors `OccupancyCount.allSupportedB`. */
export function supported(minMembers: number, clusters: Clustering): boolean {
  return clusters.every((g) => supportedCluster(minMembers, g));
}

/* ------------------------------------------------------------------ */
/*  Runtime promotion check (epistemic outcome)                       */
/* ------------------------------------------------------------------ */

/**
 * Epistemic outcome of an occupancy-count observation. Identical vocabulary to
 * the sibling mesh carriers, kept as a local literal union so this certificate
 * has no import coupling:
 *
 *  - `claim_admitted`    — every occupant cluster is tight, the occupants are
 *                          mutually well-separated, and every cluster clears the
 *                          modulation support floor, so the occupancy count (the
 *                          clustering length) may be promoted.
 *  - `observation_only`  — the empty room: no modulation above baseline, no
 *                          occupant clusters, so the honest count is 0 — reported,
 *                          never promoted (there is no count to certify or refute).
 *  - `claim_withheld`    — a gate failed (an over-merge that is not tight, or an
 *                          over-split whose reps are too close, or an unsupported
 *                          sub-floor fragment), so the proposed count is refused.
 */
export type OccupancyEpistemicOutcome =
  | 'claim_admitted'
  | 'observation_only'
  | 'claim_withheld';

/** The reason an occupancy-count claim was held back from promotion. */
export type OccupancyLimitationClass =
  | 'over_merge_undercount'
  | 'over_split_overcount'
  | 'subfloor_fragment'
  | 'empty_room'
  | 'none';

export interface OccupancyAdmission {
  readonly outcome: OccupancyEpistemicOutcome;
  readonly limitation: OccupancyLimitationClass;
  /** The proposed occupancy count (the clustering length, displayable). */
  readonly count: number;
  /** Whether every cluster passed the tightness gate. */
  readonly tight: boolean;
  /** Whether the clusters passed the separation gate. */
  readonly separated: boolean;
  /** Whether every cluster passed the support gate. */
  readonly supported: boolean;
  readonly summary: string;
}

export interface OccupancyAdmitOptions {
  /** Max member spread to a cluster's representative (tightness gate). */
  readonly intraTol: number;
  /** Reps must differ by more than this (separation gate). */
  readonly sepThreshold: number;
  /** Minimum members for a cluster to count as a person (support gate). */
  readonly minMembers: number;
}

/**
 * The promotion check Monster Mesh runs before any UI surface displays an
 * occupancy count. It folds the Lean contract into an epistemic outcome:
 *
 *  - the tightness gate refutes an over-merge: a cluster smearing two occupants
 *    together is not tight, so the under-count is withheld (mirroring
 *    `over_merge_rejected`);
 *  - the separation gate refutes an over-split: representatives within
 *    `sepThreshold` carve one occupant into phantoms, so the over-count is
 *    withheld;
 *  - the support gate refutes a sub-floor fragment: a single-member cluster is
 *    carrier noise, not a person (mirroring `subfloor_rejected`);
 *  - only a tight, separated, supported clustering is admitted, with the count
 *    set to the clustering length (mirroring `two_people_admitted` /
 *    `two_people_count`).
 *
 * The empty room is the degenerate observation-only case: an honest count of 0,
 * no one present, nothing to certify or refute (mirroring `empty_room_zero`).
 *
 * @param clusters the proposed occupancy clustering (the occupant partition)
 * @param opts      the three gate thresholds
 */
export function admitOccupancy(
  clusters: Clustering,
  opts: OccupancyAdmitOptions,
): OccupancyAdmission {
  const count = clusters.length;
  const isTight = allTight(opts.intraTol, clusters);
  const isSeparated = wellSeparated(opts.sepThreshold, clusters);
  const isSupported = supported(opts.minMembers, clusters);

  if (count === 0) {
    return {
      outcome: 'observation_only',
      limitation: 'empty_room',
      count,
      tight: isTight,
      separated: isSeparated,
      supported: isSupported,
      summary: 'empty room: no modulation above baseline, honest occupancy count 0',
    };
  }

  // Tightness gate: an over-merge crams well-separated occupants into one cluster
  // whose internal spread exceeds intraTol — refuse the under-count.
  if (!isTight) {
    return {
      outcome: 'claim_withheld',
      limitation: 'over_merge_undercount',
      count,
      tight: isTight,
      separated: isSeparated,
      supported: isSupported,
      summary: `over-merge: a cluster's internal spread exceeds intraTol ${opts.intraTol} — proposed count ${count} under-counts the occupants`,
    };
  }

  // Separation gate: an over-split carves one occupant's jitter into fragments
  // whose reps sit within sepThreshold — refuse the over-count.
  if (!isSeparated) {
    return {
      outcome: 'claim_withheld',
      limitation: 'over_split_overcount',
      count,
      tight: isTight,
      separated: isSeparated,
      supported: isSupported,
      summary: `over-split: two cluster representatives are within sepThreshold ${opts.sepThreshold} — proposed count ${count} over-counts the occupants`,
    };
  }

  // Support gate: every cluster must clear the modulation floor to be a person.
  if (!isSupported) {
    return {
      outcome: 'claim_withheld',
      limitation: 'subfloor_fragment',
      count,
      tight: isTight,
      separated: isSeparated,
      supported: isSupported,
      summary: `sub-floor: a cluster has fewer than minMembers ${opts.minMembers} observations — carrier noise, not a person`,
    };
  }

  return {
    outcome: 'claim_admitted',
    limitation: 'none',
    count,
    tight: isTight,
    separated: isSeparated,
    supported: isSupported,
    summary: `tight, well-separated, supported clustering — occupancy count ${count} certified`,
  };
}

/* ------------------------------------------------------------------ */
/*  Concrete fixtures (mirror the Lean fixtures)                      */
/* ------------------------------------------------------------------ */

/** The modulation window: three observations near 20-22 (one occupant's
 *  breathing-band cluster) and two near 60-61 (a second occupant).
 *  Mirrors `OccupancyCount.window`. */
export const window: Window = [20, 21, 22, 60, 61];

/** Intra-cluster tolerance (max member spread to representative). */
export const intraTol = 5;

/** Inter-cluster separation threshold (occupant reps must differ by more than this). */
export const sepThreshold = 10;

/** Modulation support floor: minimum members for a cluster to count as a person. */
export const minMembers = 2;

/** The correct 2-occupant grouping of the window: two tight, well-separated,
 *  supported clusters. Admitted count 2. Mirrors `OccupancyCount.twoPeople`. */
export const twoPeople: Clustering = [
  [20, 21, 22],
  [60, 61],
];

/** The over-merged 1-occupant grouping of the SAME window: both occupants crammed
 *  into one cluster whose internal spread (41) blows past `intraTol`. Rejected.
 *  Mirrors `OccupancyCount.mergedOne`. */
export const mergedOne: Clustering = [[20, 21, 22, 60, 61]];

/** The empty room: no modulation above baseline, no occupant clusters.
 *  observation_only with count 0. Mirrors `OccupancyCount.emptyRoom`. */
export const emptyRoom: Clustering = [];

/** A lone sub-floor cluster: a single modulation perturbation that never cleared
 *  the support floor (`minMembers = 2`). Carrier noise, not a person → rejected.
 *  Mirrors `OccupancyCount.subfloorOne`. */
export const subfloorOne: Clustering = [[30]];

/**
 * Occupancy Count master witness. Bundles the boundary facts the Lean theorem
 * `occupancy_count_master` proves, each recomputed here from first principles
 * (never imported from Lean IR).
 *
 * Mirrors `OccupancyCount.occupancy_count_master`.
 */
export const OccupancyMaster = {
  /** The two honest occupant clusters are each internally tight (spreads 2, 1 <= 5). */
  twoPeopleTight: allTight(intraTol, twoPeople) === true,

  /** The merged cluster is NOT tight: its spread (41) exceeds `intraTol`. */
  mergedOneNotTight: allTight(intraTol, mergedOne) === false,

  /** The two honest occupant clusters are well-separated (reps 20 vs 60, gap 40 > 10). */
  twoPeopleSeparated: wellSeparated(sepThreshold, twoPeople) === true,

  /** Every honest occupant cluster clears the support floor (lengths 3,2 >= 2). */
  twoPeopleSupported: supported(minMembers, twoPeople) === true,

  /** The lone sub-floor cluster fails the support floor (length 1 < 2). */
  subfloorOneUnsupported: supported(minMembers, subfloorOne) === false,

  /** The correct 2-occupant grouping is admitted. */
  twoPeopleAdmitted:
    admitOccupancy(twoPeople, { intraTol, sepThreshold, minMembers }).outcome ===
    'claim_admitted',

  /** The admitted occupancy count for the honest grouping is 2. */
  twoPeopleCount:
    admitOccupancy(twoPeople, { intraTol, sepThreshold, minMembers }).count === 2,

  /** The empty room is observation_only with count 0. */
  emptyRoomObservationOnly:
    admitOccupancy(emptyRoom, { intraTol, sepThreshold, minMembers }).outcome ===
    'observation_only',

  /** The empty room's count is 0. */
  emptyRoomCount:
    admitOccupancy(emptyRoom, { intraTol, sepThreshold, minMembers }).count === 0,

  /** Over-merge rejection: the 1-cluster proposal fails the tightness gate
   *  (under-count). */
  overMergeRejected:
    admitOccupancy(mergedOne, { intraTol, sepThreshold, minMembers }).outcome ===
    'claim_withheld',

  /** Sub-floor rejection: a lone single-member cluster fails the support floor. */
  subfloorRejected:
    admitOccupancy(subfloorOne, { intraTol, sepThreshold, minMembers })
      .outcome === 'claim_withheld',

  /** The hard negative: ONE window, two competing counts — count 2 admitted, the
   *  over-merged count 1 over the same window rejected. Separation fixes the count. */
  countFixedBySeparation:
    admitOccupancy(twoPeople, { intraTol, sepThreshold, minMembers }).outcome ===
      'claim_admitted' &&
    admitOccupancy(mergedOne, { intraTol, sepThreshold, minMembers }).outcome ===
      'claim_withheld',
} as const;
