// @a0n/gnosis/distributed-inference-host/src/openai-server-mesh.ts
import { Pipeline, type PipelineConfig, type PipelineTerminalProsodyConfig } from './pipeline';
import { openStation } from './transport-selector';
import { Station, type StationEndpoint } from './station';
import type { IStation } from './types';

export interface OpenAiServerMeshConfig {
  stationUrls: string[];
  layerRanges: Array<[number, number]>;
  hiddenDim: number;
  vocabSize?: number;
  chunkSize?: number;
  cobordismUrls?: string[];
  cobordismRanges?: Array<[number, number]>;
}

export interface OpenAiServerTopology {
  kind: 'mesh';
  stationCount: number;
  cobordismShardCount: number;
  layerRanges: Array<[number, number]>;
  stationUrls: string[];
}

export interface BuiltOpenAiServerMesh {
  pipeline: Pipeline;
  topology: OpenAiServerTopology;
}

export interface BuildOpenAiServerMeshOptions {
  stopTokens: number[];
  terminalProsody?: PipelineTerminalProsodyConfig;
  /** Model identity → domain-separates the Amplituhedron volume-cache key. */
  modelId?: string;
}

type EnvLike = Record<string, string | undefined>;

function splitCsv(raw: string | undefined): string[] {
  if (raw == null || raw.trim() === '') return [];
  return raw
    .split(',')
    .map((piece) => piece.trim())
    .filter((piece) => piece.length > 0);
}

function parsePositiveInteger(raw: string | undefined, name: string): number | undefined {
  if (raw == null || raw.trim() === '') return undefined;
  const parsed = Number.parseInt(raw.trim(), 10);
  if (!Number.isInteger(parsed) || parsed <= 0) {
    throw new Error(`${name} must be a positive integer`);
  }
  return parsed;
}

/**
 * Parses the Open Ai Server Layer Ranges.
 */
export function parseOpenAiServerLayerRanges(
  raw: string | undefined,
  name = 'MESH_LAYER_RANGES',
): Array<[number, number]> {
  const ranges = splitCsv(raw).map((piece) => {
    const match = piece.match(/^(\d+)\s*[:-]\s*(\d+)$/);
    if (!match) {
      throw new Error(`${name}: invalid range "${piece}" (expected start:end)`);
    }
    const start = Number.parseInt(match[1]!, 10);
    const end = Number.parseInt(match[2]!, 10);
    if (!Number.isInteger(start) || !Number.isInteger(end) || end <= start) {
      throw new Error(`${name}: invalid range "${piece}" (end must be greater than start)`);
    }
    return [start, end] as [number, number];
  });
  if (ranges.length === 0) {
    throw new Error(`${name} must contain at least one range`);
  }
  return ranges;
}

/**
 * Resolves the Open Ai Server Mesh Config.
 */
export function resolveOpenAiServerMeshConfig(
  env: EnvLike = process.env,
): OpenAiServerMeshConfig | null {
  const stationUrls = splitCsv(env['MESH_STATION_URLS']);
  if (stationUrls.length === 0) return null;

  const layerRanges = parseOpenAiServerLayerRanges(env['MESH_LAYER_RANGES']);
  if (stationUrls.length !== layerRanges.length) {
    throw new Error(
      `MESH_STATION_URLS has ${stationUrls.length} entries but MESH_LAYER_RANGES has ${layerRanges.length}`,
    );
  }

  const hiddenDim = parsePositiveInteger(env['MESH_HIDDEN_DIM'], 'MESH_HIDDEN_DIM');
  if (hiddenDim === undefined) {
    throw new Error('MESH_HIDDEN_DIM is required when MESH_STATION_URLS is set');
  }

  const vocabSize = parsePositiveInteger(env['MESH_VOCAB_SIZE'], 'MESH_VOCAB_SIZE');
  const chunkSize = parsePositiveInteger(env['MESH_CHUNK_SIZE'], 'MESH_CHUNK_SIZE');

  const cobordismUrls = splitCsv(env['MESH_COBORDISM_URLS']);
  const cobordismRangesRaw = env['MESH_COBORDISM_RANGES'];
  let cobordismRanges: Array<[number, number]> | undefined;
  if (cobordismUrls.length > 0 || (cobordismRangesRaw != null && cobordismRangesRaw.trim() !== '')) {
    if (cobordismUrls.length === 0) {
      throw new Error('MESH_COBORDISM_URLS is required when MESH_COBORDISM_RANGES is set');
    }
    cobordismRanges = parseOpenAiServerLayerRanges(
      cobordismRangesRaw,
      'MESH_COBORDISM_RANGES',
    );
    if (cobordismUrls.length !== cobordismRanges.length) {
      throw new Error(
        `MESH_COBORDISM_URLS has ${cobordismUrls.length} entries but MESH_COBORDISM_RANGES has ${cobordismRanges.length}`,
      );
    }
  }

  const config: OpenAiServerMeshConfig = {
    stationUrls,
    layerRanges,
    hiddenDim,
  };
  if (vocabSize !== undefined) config.vocabSize = vocabSize;
  if (chunkSize !== undefined) config.chunkSize = chunkSize;
  if (cobordismUrls.length > 0) config.cobordismUrls = cobordismUrls;
  if (cobordismRanges !== undefined) config.cobordismRanges = cobordismRanges;
  return config;
}

function stationRole(index: number, total: number): 'entry' | 'mid' | 'exit' | undefined {
  if (total === 1) return undefined;
  if (index === 0) return 'entry';
  if (index === total - 1) return 'exit';
  return 'mid';
}

function createSerializedHttpEndpoint(): StationEndpoint {
  let mutex: Promise<unknown> = Promise.resolve();
  return {
    fetch(req: Request): Promise<Response> {
      const run = mutex.then(() => fetch(req), () => fetch(req));
      mutex = run.then(
        () => undefined,
        () => undefined,
      );
      return run;
    },
  };
}

function localStationMaxBatchSize(): number {
  const raw = process.env['MESH_LOCAL_STATION_MAX_BATCH'] ?? process.env['MOONSHINE_PREFILL_BATCH'];
  const parsed = raw ? Number.parseInt(raw, 10) : 128;
  if (!Number.isInteger(parsed) || parsed <= 0) return 128;
  return Math.min(128, parsed);
}

/**
 * Builds the Open Ai Server Mesh Pipeline.
 */
export function buildOpenAiServerMeshPipeline(
  config: OpenAiServerMeshConfig,
  options: BuildOpenAiServerMeshOptions,
): BuiltOpenAiServerMesh {
  const urlCounts = new Map<string, number>();
  for (const url of config.stationUrls) {
    urlCounts.set(url, (urlCounts.get(url) ?? 0) + 1);
  }
  const serializedHttpEndpoints = new Map<string, StationEndpoint>();
  const stations: IStation[] = config.stationUrls.map((url, index) => {
    const layerRange = config.layerRanges[index]!;
    const role = stationRole(index, config.stationUrls.length);
    if ((url.startsWith('http://') || url.startsWith('https://')) && (urlCounts.get(url) ?? 0) > 1) {
      let endpoint = serializedHttpEndpoints.get(url);
      if (!endpoint) {
        endpoint = createSerializedHttpEndpoint();
        serializedHttpEndpoints.set(url, endpoint);
      }
      const [startLayer, endLayer] = layerRange;
      const stationConfig = {
        endpoint,
        baseUrl: url,
        startLayer,
        endLayer,
        maxBatchSize: localStationMaxBatchSize(),
        ...(role !== undefined ? { role } : {}),
      };
      return new Station(stationConfig);
    }
    return openStation({
      url,
      layerRange,
      ...(role !== undefined ? { role } : {}),
    });
  });

  const pipelineConfig: PipelineConfig = {
    stations,
    hiddenDim: config.hiddenDim,
    stopTokens: options.stopTokens,
  };
  if (options.modelId !== undefined) pipelineConfig.amplituhedronModelKey = options.modelId;
  if (config.vocabSize !== undefined) pipelineConfig.vocabSize = config.vocabSize;
  if (config.chunkSize !== undefined) pipelineConfig.chunkSize = config.chunkSize;
  if (options.terminalProsody !== undefined) {
    pipelineConfig.terminalProsody = options.terminalProsody;
  }

  if (config.cobordismUrls && config.cobordismRanges) {
    pipelineConfig.cobordismStations = config.cobordismUrls.map((url, index) => ({
      station: openStation({
        url,
        layerRange: [0, 0],
        role: 'cobordism',
      }),
      vocabStart: config.cobordismRanges![index]![0],
      vocabEnd: config.cobordismRanges![index]![1],
    }));
  }

  return {
    pipeline: new Pipeline(pipelineConfig),
    topology: {
      kind: 'mesh',
      stationCount: stations.length,
      cobordismShardCount: pipelineConfig.cobordismStations?.length ?? 0,
      layerRanges: config.layerRanges.map((range) => [range[0], range[1]]),
      stationUrls: config.stationUrls.slice(),
    },
  };
}
