// @a0n/gnosis/distributed-inference-host/src/particle-physics.ts
/**
 * # Particle physics — runtime mirror of the Standard-Model arc
 *
 * Mirrors the Lean modules:
 * - `Gnosis.BosonSkyrmsEquilibria` (already in `boson-skyrms.ts`)
 * - `Gnosis.FermionExclusionEquilibria`
 * - `Gnosis.CompositeParticles`
 * - `Gnosis.DarkSectorEquilibria`
 * - `Gnosis.PhaseTransitionLadder`
 * - `Gnosis.SuperstringDimensionDerivation`
 * - `Gnosis.CostAlgebraDimensionNoGo`
 * - `Gnosis.NahmDimensionCeiling`
 *
 * The runtime exposes the catalogs and predicates so an inference
 * mesh tick can classify carriers as boson/fermion/composite/dark
 * equilibria, and so the dimensional sandwich (10 ≤ d ≤ 11) is
 * available as a runtime invariant.
 *
 * All claims here are *signature*, not *derivation*: the cost-algebra
 * permits arbitrary phase counts via `every_phase_count_is_a_tower`.
 * Naming the SM walls is decoration. The Higgs ↔ vacuum identity and
 * the Pauli ⇔ no-cloning identity are the formal payoffs.
 */

import { TOWER_CEILINGS, type BuleyProjection } from './spectral-noise-monitor';

// ────────────────────────────────────────────────────────────────────
// Fermion exclusion equilibria
// ────────────────────────────────────────────────────────────────────

export type Generation = 'first' | 'second' | 'third';
export type FermionFlavor =
  | 'quarkUp'
  | 'quarkDown'
  | 'leptonCharged'
  | 'leptonNeutrino';
export type Antiparticle = 'particle' | 'antiparticle';

export interface StandardModelFermion {
  generation: Generation;
  flavor: FermionFlavor;
  side: Antiparticle;
}

/** All 24 SM fermions. */
export const allStandardModelFermions: readonly StandardModelFermion[] =
  (['first', 'second', 'third'] as const).flatMap((generation) =>
    (
      ['quarkUp', 'quarkDown', 'leptonCharged', 'leptonNeutrino'] as const
    ).flatMap((flavor) =>
      (['particle', 'antiparticle'] as const).map((side) => ({
        generation,
        flavor,
        side,
      })),
    ),
  );

/** All fermions share the Dodecagon ceiling (12). */
export function fermionPhaseCount(_f: StandardModelFermion): number {
  return TOWER_CEILINGS.dodecagon;
}

/**
 * Handles the gnosis fermion Skyrms Equilibrium workflow.
 */
export function fermionSkyrmsEquilibrium(
  b: BuleyProjection,
  f: StandardModelFermion,
): boolean {
  return b.score === fermionPhaseCount(f);
}

/** Pauli exclusion as no-cloning: cloning a fermion at a positive
 *  score generates entropy equal to its score. The diagonal is not a
 *  CostHom unless the carrier is the vacuum. */
export function fermionPauliExclusionEntropy(b: BuleyProjection): number {
  return b.score; // entropy generated by cloning
}

/** A score-positive fermion always pays the Pauli tax. */
export function fermionPaysExclusionTax(b: BuleyProjection): boolean {
  return b.score > 0;
}

// ────────────────────────────────────────────────────────────────────
// Composite particles (mesons, baryons via collide)
// ────────────────────────────────────────────────────────────────────

export interface CompositeParticle {
  kind: 'meson' | 'baryon';
  constituents: readonly BuleyProjection[];
  /** Total energy = sum of constituent scores (conservation). */
  totalEnergy: number;
}

/** A meson: quark + antiquark. */
export function meson(q: BuleyProjection, qbar: BuleyProjection): CompositeParticle {
  return {
    kind: 'meson',
    constituents: [q, qbar],
    totalEnergy: q.score + qbar.score,
  };
}

/** A baryon: three quarks. */
export function baryon(
  q1: BuleyProjection,
  q2: BuleyProjection,
  q3: BuleyProjection,
): CompositeParticle {
  return {
    kind: 'baryon',
    constituents: [q1, q2, q3],
    totalEnergy: q1.score + q2.score + q3.score,
  };
}

// ────────────────────────────────────────────────────────────────────
// Dark sector equilibria
// ────────────────────────────────────────────────────────────────────

export type DarkSectorParticle =
  | { kind: 'hexon'; index: number /* 0..5 */ }
  | { kind: 'pentagon' }
  | { kind: 'septagon' }
  | { kind: 'decagon' }
  | { kind: 'hendecagon' };

/**
 * Handles the gnosis dark Phase Count workflow.
 */
export function darkPhaseCount(d: DarkSectorParticle): number {
  switch (d.kind) {
    case 'hexon':
      return TOWER_CEILINGS.hexon; // 6
    case 'pentagon':
      return 5;
    case 'septagon':
      return 7;
    case 'decagon':
      return TOWER_CEILINGS.decagon; // 10
    case 'hendecagon':
      return TOWER_CEILINGS.hendecagon; // 11
  }
}

/**
 * Handles the gnosis dark Skyrms Equilibrium workflow.
 */
export function darkSkyrmsEquilibrium(
  b: BuleyProjection,
  d: DarkSectorParticle,
): boolean {
  return b.score === darkPhaseCount(d);
}

/** SM walls are 0 (Higgs), 3 (electroweak), 8 (gluon), 12 (fermion).
 *  Dark walls don't intersect any of these. */
export function darkDoesNotIntersectSM(d: DarkSectorParticle): boolean {
  const phase = darkPhaseCount(d);
  return phase !== 0 && phase !== 3 && phase !== 8 && phase !== 12;
}

// ────────────────────────────────────────────────────────────────────
// Phase-transition ladder + dimension sandwich
// ────────────────────────────────────────────────────────────────────

/**
 * Handles the gnosis phase Transition Distance workflow.
 */
export function phaseTransitionDistance(m: number, n: number): number {
  return m <= n ? n - m : m - n;
}

/** Cost-algebra minimum-generator dimensions. */
export const COST_ALGEBRA_AXES = {
  buleFaces: 3,
  biSided: 2,
  temporal: 3,
  vacuum: 1,
  clinamen: 1,
} as const;

/** Sum of forced minimums = 10 (superstring). */
export const SUPERSTRING_DIMENSION =
  COST_ALGEBRA_AXES.buleFaces +
  COST_ALGEBRA_AXES.biSided +
  COST_ALGEBRA_AXES.temporal +
  COST_ALGEBRA_AXES.vacuum +
  COST_ALGEBRA_AXES.clinamen;

/** M-theory = superstring + gauge orientation (one clinamen). */
export const M_THEORY_DIMENSION = SUPERSTRING_DIMENSION + 1;

/** Bosonic-string = superstring + doubled-Octagon (16). */
export const BOSONIC_STRING_DIMENSION = SUPERSTRING_DIMENSION + 16;

/** The clinamen step IS the coupling constant — one quantum. */
export const COUPLING_CONSTANT = 1;

/** Witten's identification: M-theory's 11th dimension is the
 *  coupling constant. Returns true if the dimension difference
 *  matches the coupling. */
export function eleventhDimensionIsCoupling(): boolean {
  return phaseTransitionDistance(SUPERSTRING_DIMENSION, M_THEORY_DIMENSION) ===
    COUPLING_CONSTANT;
}

/** A Nahm-minimal axis-set check. */
export interface NahmMinimalAxisSet {
  buleFaces: number;
  biSided: number;
  temporal: number;
  vacuum: number;
  clinamen: number;
  gauge: number;
  doubledOctagon: number;
}

/** Predicate: is this axis set in the Nahm-minimal regime?
 *  (Each axis at its forced cardinality, gauge ≤ 1, no doubled-Octagon.) */
export function isNahmMinimal(a: NahmMinimalAxisSet): boolean {
  return (
    a.buleFaces === 3 &&
    a.biSided === 2 &&
    a.temporal === 3 &&
    a.vacuum === 1 &&
    a.clinamen === 1 &&
    a.gauge <= 1 &&
    a.doubledOctagon === 0
  );
}

/**
 * Handles the gnosis axis Set Total workflow.
 */
export function axisSetTotal(a: NahmMinimalAxisSet): number {
  return (
    a.buleFaces +
    a.biSided +
    a.temporal +
    a.vacuum +
    a.clinamen +
    a.gauge +
    a.doubledOctagon
  );
}

/** Nahm sandwich: every Nahm-minimal axis set has dimension exactly
 *  10 or 11. */
export function nahmSandwichDimension(a: NahmMinimalAxisSet): 10 | 11 | null {
  if (!isNahmMinimal(a)) return null;
  const total = axisSetTotal(a);
  if (total === 10) return 10;
  if (total === 11) return 11;
  return null;
}
