// @a0n/gnosis/distributed-inference-host/src/personality-orthography.ts
/**
 * personality-orthography.ts
 * ==========================
 *
 * TS mirror of the orthographic scorer in
 * `open-source/gnosis/distributed-inference/src/personality_steering.rs`
 * (itself bounded by `Gnosis.PersonalityLogitSteering`).
 *
 * The fat-station operates on token *ids*; the model's tokenizer (token strings)
 * lives here in the openai-server. So the orthographic compile — which needs the
 * surface text of each token to detect caps / punctuation / emoji / emoticons —
 * runs in TS, producing a per-token signed bias vector (`boundary_bias`) that is
 * installed on the fat-station via the precompiled Path-A install form. The math
 * is a 1:1 mirror of the Rust/Lean contract, pinned by the parity tests.
 */

import type {
  PersonalityStylePreferences,
  PersonalitySteeringSignature,
} from './types';

export interface TokenStyleFeatures {
  allCaps: boolean;
  allLower: boolean;
  hasExclaim: boolean;
  hasQuestion: boolean;
  hasEllipsis: boolean;
  hasTerminalPeriod: boolean;
  hasEmoji: boolean;
  hasAsciiEmoticon: boolean;
}

const UPPER = /\p{Lu}/u;
const LOWER = /\p{Ll}/u;
const LETTER = /\p{L}/u;

function isEmojiChar(c: string): boolean {
  const u = c.codePointAt(0) ?? 0;
  return (
    (u >= 0x1f300 && u <= 0x1faff) ||
    (u >= 0x2600 && u <= 0x27bf) ||
    (u >= 0x1f1e6 && u <= 0x1f1ff) ||
    u === 0x2764 ||
    u === 0xfe0f
  );
}

const EMOTICONS = [
  ':)', ':-)', ':(', ':-(', ';)', ';-)', ':d', ':-d', ':p', ':-p', '<3', '^^',
];

function containsAsciiEmoticon(tok: string): boolean {
  const lower = tok.toLowerCase();
  return EMOTICONS.some((e) => lower.includes(e));
}

/**
 * Converts input into token Style Features.
 */
export function tokenStyleFeatures(tok: string): TokenStyleFeatures {
  const chars = [...tok];
  const letters = chars.filter((c) => LETTER.test(c));
  const hasLetters = letters.length > 0;
  const anyUpper = letters.some((c) => UPPER.test(c));
  const anyLower = letters.some((c) => LOWER.test(c));
  const allCaps = letters.length >= 2 && anyUpper && !anyLower;
  const allLower = hasLetters && !anyUpper;

  const hasEllipsis = tok.includes('…') || tok.includes('...');
  const hasQuestion = tok.includes('?');
  const hasExclaim = tok.includes('!');
  const hasTerminalPeriod = tok.trimEnd().endsWith('.') && !hasEllipsis;
  const hasEmoji = chars.some(isEmojiChar);
  const hasAsciiEmoticon = containsAsciiEmoticon(tok);

  return {
    allCaps,
    allLower,
    hasExclaim,
    hasQuestion,
    hasEllipsis,
    hasTerminalPeriod,
    hasEmoji,
    hasAsciiEmoticon,
  };
}

export interface SignedAxisAffinity {
  behavioral: number;
  aesthetic: number;
  prosody: number;
  harmonic: number;
}

const n = (x: number | undefined): number =>
  typeof x === 'number' && Number.isFinite(x) ? Math.trunc(x) : 0;

/** Mirror of Rust `orthographic_affinity` (signed, per-axis; integer halves). */
export function orthographicAffinity(
  tok: string,
  prefs: PersonalityStylePreferences,
): SignedAxisAffinity {
  const f = tokenStyleFeatures(tok);
  const a: SignedAxisAffinity = {
    behavioral: 0,
    aesthetic: 0,
    prosody: 0,
    harmonic: 0,
  };
  if (f.allCaps) {
    a.behavioral += n(prefs.caps);
    a.aesthetic += Math.trunc(n(prefs.caps) / 2);
  }
  if (f.allLower) a.aesthetic += n(prefs.lower);
  if (f.hasExclaim) a.behavioral += n(prefs.exclaim);
  if (f.hasQuestion) a.prosody += n(prefs.question);
  if (f.hasEllipsis) a.prosody += n(prefs.ellipsis);
  if (f.hasTerminalPeriod) a.prosody += n(prefs.period);
  if (f.hasEmoji) {
    a.aesthetic += n(prefs.emoji);
    a.behavioral += Math.trunc(n(prefs.emoji) / 2);
  }
  if (f.hasAsciiEmoticon) {
    a.aesthetic += n(prefs.ascii_emoticon);
    a.behavioral += Math.trunc(n(prefs.ascii_emoticon) / 2);
  }
  return a;
}

/** Mirror of Rust `fold_signed`. */
export function foldSigned(
  sig: PersonalitySteeringSignature,
  a: SignedAxisAffinity,
): number {
  return (
    n(sig.behavioral_gain) * a.behavioral +
    n(sig.aesthetic_gain) * a.aesthetic +
    n(sig.prosody_gain) * a.prosody +
    n(sig.harmonic_gain) * a.harmonic
  );
}

/** Mirror of Lean/Rust `clampSym`: symmetric clamp into [-cap, +cap]. */
export function clampSym(x: number, cap: number): number {
  const c = Math.max(0, n(cap));
  if (x > c) return c;
  if (x < -c) return -c;
  return x;
}

/**
 * Compile the per-token signed bias vector from a steering spec and the model's
 * vocab strings (vocab[id] = surface text of token id). Each entry is
 * `clampSym(fold_signed(orthographic_affinity(tok)), steer_cap)` — the Path-A
 * `boundary_bias` the fat-station applies directly before the argmax.
 */
export function compileBoundaryBias(
  prefs: PersonalityStylePreferences,
  sig: PersonalitySteeringSignature,
  vocab: readonly string[],
): number[] {
  const cap = n(sig.steer_cap);
  return vocab.map((tok) => clampSym(foldSigned(sig, orthographicAffinity(tok, prefs)), cap));
}
