// @a0n/gnosis/distributed-inference-host/src/phonemizer.ts
/**
 * Pure-TypeScript English G2P (grapheme-to-phoneme) front-end for Pneuma TTS.
 *
 * Maps clean English text → phoneme IDs that the Pneuma TTS acoustic model
 * (`model_pneuma_tts_acoustic.rs`) consumes. v1 is a hand-rolled rule-based
 * tokenizer with ~70 letter / digraph / suffix rules — enough to make
 * "the quick brown fox" produce intelligible audio. Phase 3 swaps this for
 * espeak-ng wasm.
 *
 * Design constraints:
 * - Zero runtime deps (pure TS, no native bindings, runs in CF Workers).
 * - Stable inventory: phoneme IDs are stable across calls so cached .knots
 *   keep the same vocabulary mapping for the lifetime of the host shim.
 * - Lifted from `open-source/gnosis/scripts/say-sentence.ts` — that script
 *   carried the production-grade heuristic ruleset that the rest of the
 *   monorepo had already iterated against.
 *
 * Coverage notes:
 * - The ARPAbet-style inventory below is ~46 phonemes + a `<pause>` and a
 *   `<unk>` slot; the remaining IDs (up to 70) are reserved for
 *   stress / boundary symbols a future phonemizer might emit.
 * - OOV behaviour: words that contain non-letter characters fall back to a
 *   per-letter spelling; pure-letter words always emit something even if the
 *   ruleset is loose.
 */

// ─────────────────────────────────────────────────────────────────────────────
// Phoneme inventory
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Phoneme inventory — index = phoneme ID. Ordering is stable; never reorder
 * (existing acoustic .knots embed phoneme IDs in their phoneme_embed table).
 *
 * Layout:
 * - 0:        `<pad>`     reserved padding / null token
 * - 1:        `<pause>`   inter-word silence (also punctuation)
 * - 2:        `<unk>`     out-of-vocabulary fallback
 * - 3..18:    vowels (16 ARPAbet vowels incl. r-coloured ER)
 * - 19..42:   consonants (24 ARPAbet consonants)
 * - 43..69:   reserved for stress markers / boundaries / future phonemes
 */
export const PHONEMES: readonly string[] = [
  // 0..2 — control tokens
  '<pad>',
  '<pause>',
  '<unk>',
  // 3..18 — vowels (ARPAbet)
  'AA', 'AE', 'AH', 'AO', 'AW', 'AY',
  'EH', 'ER', 'EY',
  'IH', 'IY',
  'OW', 'OY',
  'UH', 'UW',
  'AX',
  // 19..42 — consonants (ARPAbet)
  'B', 'CH', 'D', 'DH', 'F', 'G', 'HH', 'JH', 'K', 'L',
  'M', 'N', 'NG', 'P', 'R', 'S', 'SH', 'T', 'TH', 'V',
  'W', 'Y', 'Z', 'ZH',
  // 43..69 — reserved (stress / boundaries / future symbols).
  // Filled with placeholder ids so PHONEMES.length is stable at 70.
  '<r43>', '<r44>', '<r45>', '<r46>', '<r47>', '<r48>', '<r49>',
  '<r50>', '<r51>', '<r52>', '<r53>', '<r54>', '<r55>', '<r56>',
  '<r57>', '<r58>', '<r59>', '<r60>', '<r61>', '<r62>', '<r63>',
  '<r64>', '<r65>', '<r66>', '<r67>', '<r68>', '<r69>',
];

/** Reverse inventory: symbol → numeric ID (built once at module load). */
const PHONEME_ID_BY_SYMBOL: Record<string, number> = (() => {
  const map: Record<string, number> = {};
  for (let i = 0; i < PHONEMES.length; i++) {
    map[PHONEMES[i]] = i;
  }
  return map;
})();

/**
 * Look up the numeric ID for a phoneme symbol.
 *
 * Returns the `<unk>` ID (2) for unknown symbols rather than throwing — the
 * phonemizer should never crash an inference call on a bad rule output.
 */
export function phonemeToId(symbol: string): number {
  const id = PHONEME_ID_BY_SYMBOL[symbol];
  return id !== undefined ? id : PHONEME_ID_BY_SYMBOL['<unk>'];
}

// ─────────────────────────────────────────────────────────────────────────────
// Heuristic G2P
// ─────────────────────────────────────────────────────────────────────────────

/** True iff `c` is one of the orthographic vowel letters (a/e/i/o/u/y). */
function isVowelChar(c: string): boolean {
  return c.length === 1 && 'aeiouy'.indexOf(c) >= 0;
}

/**
 * Detect a "silent E" pattern at the end of `word`: a final `e` (or `es`)
 * preceded by C-V (consonant-vowel) — the classic English rule that
 * lengthens the prior vowel ("hat" vs "hate", "win" vs "wine").
 */
function hasSilentE(word: string): boolean {
  const len = word.length;
  if (len > 2 && word.endsWith('e') && !isVowelChar(word[len - 2]) && isVowelChar(word[len - 3])) {
    return true;
  }
  if (
    len > 3
    && word.endsWith('es')
    && !isVowelChar(word[len - 3])
    && isVowelChar(word[len - 4])
  ) {
    return true;
  }
  return false;
}

/**
 * Convert a single lowercase word (a-z only) to a sequence of phoneme symbols.
 *
 * Pattern lifted from `gnosis/scripts/say-sentence.ts` — same precedence:
 * suffixes → digraphs / vowel clusters → consonant digraphs → vowel+r
 * (r-coloured) → consonants → individual vowels → fallback.
 */
function wordToPhonemesHeuristic(word: string): string[] {
  const result: string[] = [];
  const len = word.length;
  if (len === 0) return result;
  const silentE = hasSilentE(word);
  let i = 0;

  while (i < len) {
    const remaining = word.slice(i);
    const ch = word[i];
    const next = word[i + 1] || '';
    const prev = i > 0 ? word[i - 1] : '';

    // 0. Suffixes (terminal logic).
    if (word.endsWith('ed') && i === len - 2 && len > 3) {
      if (prev === 't' || prev === 'd') {
        result.push('AH', 'D');
      } else if ('pkfshcx'.indexOf(prev) >= 0) {
        result.push('T');
      } else {
        result.push('D');
      }
      i += 2;
      continue;
    }

    // 1. Digraphs and common vowel clusters.
    if (remaining.startsWith('through')) { result.push('TH', 'R', 'UW'); i += 7; continue; }
    if (remaining.startsWith('tion')) { result.push('SH', 'AH', 'N'); i += 4; continue; }
    if (remaining.startsWith('sion')) {
      result.push(isVowelChar(prev) ? 'ZH' : 'SH', 'AH', 'N');
      i += 4; continue;
    }
    if (remaining.startsWith('th')) { result.push('DH'); i += 2; continue; }
    if (remaining.startsWith('sh')) { result.push('SH'); i += 2; continue; }
    if (remaining.startsWith('ch')) { result.push('CH'); i += 2; continue; }
    if (remaining.startsWith('ight')) { result.push('AY', 'T'); i += 4; continue; }
    if (remaining.startsWith('ook')) { result.push('UH', 'K'); i += 3; continue; }
    if (remaining.startsWith('ing')) { result.push('IH', 'NG'); i += 3; continue; }
    if (remaining.startsWith('ng')) { result.push('NG'); i += 2; continue; }
    if (remaining.startsWith('ear')) { result.push('IY', 'R'); i += 3; continue; }
    if (remaining.startsWith('air')) { result.push('EH', 'R'); i += 3; continue; }
    if (remaining.startsWith('ee') || remaining.startsWith('ea')) { result.push('IY'); i += 2; continue; }
    if (remaining.startsWith('oo')) { result.push('UW'); i += 2; continue; }
    if (remaining.startsWith('ay') || remaining.startsWith('ai')) { result.push('EY'); i += 2; continue; }

    if (remaining.startsWith('ou') || remaining.startsWith('ow')) {
      // Structural attractor: 'AW' (out, down, cow) before alveolar/lateral
      // closures, 'OW' otherwise (show, soul).
      const followingChar = word[i + 2] || '';
      if ('ntdl'.indexOf(followingChar) >= 0) {
        result.push('AW');
      } else {
        result.push('OW');
      }
      i += 2;
      continue;
    }

    // 1.5 Consonant digraphs and silent clusters.
    if (remaining.startsWith('ph')) { result.push('F'); i += 2; continue; }
    if (remaining.startsWith('kn') && i === 0) { result.push('N'); i += 2; continue; }
    if (remaining.startsWith('wr') && i === 0) { result.push('R'); i += 2; continue; }
    if (remaining.startsWith('wh')) { result.push('W'); i += 2; continue; }
    if (remaining.startsWith('gh')) {
      if (i === 0) {
        result.push('G');
      } else if (prev === 'u' && (word[i - 2] === 'a' || word[i - 2] === 'o')) {
        if (i + 2 === len) result.push('F');
      }
      i += 2;
      continue;
    }

    // 2. Vowel + R (r-coloured vowels).
    const isTerminal = (i + 2 === len) || (i + 3 === len && word[i + 2] === 's');

    if (ch === 'a' && next === 'r') {
      if (silentE && i + 2 === len - 1) { result.push('EH', 'R'); i += 2; continue; }
      result.push('AA', 'R'); i += 2; continue;
    }
    if (ch === 'o' && next === 'r') {
      if (isTerminal) { result.push('ER'); i += 2; continue; }
      result.push('AO'); i += 1; continue;
    }
    if (ch === 'e' && next === 'r') { result.push('ER'); i += 2; continue; }
    if (ch === 'i' && next === 'r') {
      if (silentE && i + 2 === len - 1) { result.push('AY', 'R'); i += 2; continue; }
      result.push('ER'); i += 2; continue;
    }
    if (ch === 'u' && next === 'r') { result.push('ER'); i += 2; continue; }

    // 3. Consonants and clusters.
    if (ch === 'c') {
      if (next === 'h') { result.push('CH'); i += 2; continue; }
      if ('eiy'.indexOf(next) >= 0) { result.push('S'); } else { result.push('K'); }
    } else if (ch === 'g') {
      if ('eiy'.indexOf(next) >= 0) { result.push('JH'); } else { result.push('G'); }
    } else if (ch === 'j') {
      result.push('JH');
    } else if (ch === 's') {
      // Intervocalic s → /Z/ ("rose", "music"); otherwise /S/.
      if (isVowelChar(prev) && isVowelChar(next)) { result.push('Z'); } else { result.push('S'); }
    } else if (ch === 'q') {
      if (next === 'u') { result.push('K', 'W'); i += 2; continue; }
      result.push('K');
    } else if (ch === 'y' && i === 0 && isVowelChar(next)) {
      result.push('Y');
    } else if (ch === 'x') {
      result.push('K', 'S');
    } else if (ch === 'b') { result.push('B'); }
    else if (ch === 'd') { result.push('D'); }
    else if (ch === 'f') { result.push('F'); }
    else if (ch === 'h') { result.push('HH'); }
    else if (ch === 'k') { result.push('K'); }
    else if (ch === 'l') { result.push('L'); }
    else if (ch === 'm') { result.push('M'); }
    else if (ch === 'n') { result.push('N'); }
    else if (ch === 'p') { result.push('P'); }
    else if (ch === 'r') { result.push('R'); }
    else if (ch === 't') { result.push('T'); }
    else if (ch === 'v') { result.push('V'); }
    else if (ch === 'w') { result.push('W'); }
    else if (ch === 'z') { result.push('Z'); }

    // 4. Individual vowels (fallback).
    else if (isVowelChar(ch)) {
      if (ch === 'e' && i === len - 1 && silentE) {
        // silent — emit nothing
      } else if (ch === 'a') {
        result.push(i === len - 1 ? 'AH' : (silentE ? 'EY' : 'AE'));
      } else if (ch === 'e') {
        result.push(i === len - 1 && len <= 3 ? 'AH' : 'EH');
      } else if (ch === 'i') {
        result.push(silentE ? 'AY' : 'IH');
      } else if (ch === 'o') {
        result.push(i === len - 1 ? 'OW' : (silentE ? 'OW' : 'AA'));
      } else if (ch === 'u') {
        result.push(silentE ? 'UW' : 'AH');
      } else if (ch === 'y') {
        result.push(len <= 3 ? 'AY' : 'IY');
      }
    }

    i++;
  }

  return result;
}

/**
 * OOV fallback: spell each letter individually, emitting per-letter
 * approximations. Used when a "word" contains characters the heuristic ruleset
 * above can't tokenize cleanly.
 */
function letterSpelling(word: string): string[] {
  const out: string[] = [];
  for (const ch of word) {
    if (ch >= 'a' && ch <= 'z') {
      out.push(...wordToPhonemesHeuristic(ch));
    }
  }
  return out;
}

// ─────────────────────────────────────────────────────────────────────────────
// Public entry point
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Tokenize raw text into a sequence of phoneme IDs.
 *
 * v1 pipeline:
 *  1. lower-case
 *  2. split on whitespace and punctuation, treating punctuation as `<pause>`
 *  3. for each word, run the heuristic G2P; OOV words fall back to letter
 *     spelling
 *  4. interleave a `<pause>` between words
 *
 * Returns: phoneme IDs `(0..PHONEMES.length)` ready to feed into the acoustic
 * model. Always returns at least one element when input is non-empty (the
 * trailing `<pause>` boundary is preserved).
 *
 * Phase 3 will replace the body with espeak-ng wasm; the signature stays
 * stable so callers don't need to change.
 */
export function textToPhonemeIds(text: string): number[] {
  if (typeof text !== 'string' || text.length === 0) return [];
  const lowered = text.toLowerCase();
  const out: number[] = [];
  const pauseId = PHONEME_ID_BY_SYMBOL['<pause>'];

  // Tokenize into runs of letters; everything else collapses to a single
  // <pause> (consecutive punctuation merges into one boundary).
  let i = 0;
  let prevWasWord = false;
  while (i < lowered.length) {
    const ch = lowered[i];
    if (ch >= 'a' && ch <= 'z') {
      // Capture a maximal letter run as a "word".
      let j = i;
      while (j < lowered.length && lowered[j] >= 'a' && lowered[j] <= 'z') j++;
      const word = lowered.slice(i, j);
      const symbols = wordToPhonemesHeuristic(word);
      const finalSymbols = symbols.length > 0 ? symbols : letterSpelling(word);
      if (prevWasWord) out.push(pauseId);
      for (const s of finalSymbols) {
        out.push(phonemeToId(s));
      }
      prevWasWord = true;
      i = j;
    } else if (ch === ' ' || ch === '\t' || ch === '\n' || ch === '\r') {
      // Whitespace: consume; pause inserted lazily before next word.
      i++;
    } else {
      // Punctuation: emit one pause and skip subsequent punctuation runs.
      if (out.length === 0 || out[out.length - 1] !== pauseId) {
        out.push(pauseId);
      }
      // Mark "no word yet" so the next letter run doesn't add a redundant pause.
      prevWasWord = false;
      i++;
    }
  }

  return out;
}
