// @a0n/gnosis/distributed-inference-host/src/phoneme-fingerprint.ts
// Per-frame 8-band phoneme energy fingerprint. Honest substitute for
// wav2vec / HMM phoneme decoders — computes fixed-band energies that
// approximate phoneme distinctiveness. Bridges to pneuma voice-flac
// when the caller wants a symbol stream without a learned model.
// Limited: distinguishes silence / voiced (vowel-formant) / fricative
// / mixed but does NOT identify specific phonemes (no /a/ vs /e/ etc.).
// The 8 bands span 50 Hz - 8 kHz, biased toward speech-relevant
// formant and fricative regions.

import { stft, type STFTConfig } from "./audio-stft";

export type PhonemeCategory = "silence" | "voiced" | "fricative" | "mixed";

export interface PhonemeFingerprint {
  /** 8 band energies (sum of |X|² over each band's freq bins). */
  bands: Float32Array;
  /** Total energy across all bands. */
  totalEnergy: number;
  /** Coarse phoneme category. */
  category: PhonemeCategory;
  /** Index of the maximum-energy band. */
  dominantBand: number;
}

export interface PhonemeFingerprintArgs {
  /** Mono audio in [-1, 1]. */
  audio: Float32Array;
  sampleRate: number;
  /** STFT FFT size. Power of 2. Default 1024. */
  nFft?: number;
  /** Hop length. Default nFft / 4. */
  hopLength?: number;
  /** Silence threshold on totalEnergy (linear, in band-energy units).
   *  Default 1e-3. Frames below this are 'silence'. */
  silenceThreshold?: number;
}

export interface PhonemeFingerprintResult {
  /** Per-frame fingerprint, length = nFrames. */
  frames: ReadonlyArray<PhonemeFingerprint>;
  /** Sample rate (echoed for caller convenience). */
  sampleRate: number;
  /** STFT config used (for downstream consumers like Griffin-Lim). */
  nFft: number;
  hopLength: number;
}

/** Per-band frequency edges in Hz, length 9 (8 bands → 9 edges). */
export const BAND_EDGES_HZ: readonly number[] = [
  50, 200, 500, 1000, 2000, 3000, 4500, 6000, 8000,
];

const NUM_BANDS = 8;

function isPowerOfTwo(n: number): boolean {
  return Number.isInteger(n) && n > 0 && (n & (n - 1)) === 0;
}

/**
 * Handles the gnosis extract Phoneme Fingerprints workflow.
 */
export function extractPhonemeFingerprints(
  args: PhonemeFingerprintArgs,
): PhonemeFingerprintResult {
  const { audio, sampleRate } = args;
  const nFft = args.nFft ?? 1024;
  const hopLength = args.hopLength ?? nFft / 4;
  const silenceThreshold = args.silenceThreshold ?? 1e-3;

  if (!(audio instanceof Float32Array)) {
    throw new RangeError("extractPhonemeFingerprints: audio must be Float32Array");
  }
  if (!Number.isFinite(sampleRate) || sampleRate <= 0) {
    throw new RangeError(
      `extractPhonemeFingerprints: sampleRate must be > 0, got ${sampleRate}`,
    );
  }
  if (!isPowerOfTwo(nFft)) {
    throw new RangeError(
      `extractPhonemeFingerprints: nFft must be a positive power of two, got ${nFft}`,
    );
  }
  if (!Number.isFinite(hopLength) || hopLength <= 0 || hopLength > nFft) {
    throw new RangeError(
      `extractPhonemeFingerprints: hopLength must be in (0, nFft], got ${hopLength}`,
    );
  }
  if (!Number.isFinite(silenceThreshold) || silenceThreshold <= 0) {
    throw new RangeError(
      `extractPhonemeFingerprints: silenceThreshold must be > 0, got ${silenceThreshold}`,
    );
  }
  if (audio.length < nFft) {
    throw new RangeError(
      `extractPhonemeFingerprints: audio length (${audio.length}) must be >= nFft (${nFft})`,
    );
  }

  const config: STFTConfig = { nFft, hopLength };
  const spec = stft(audio, config);
  const { magnitude, nFrames, nFreqs } = spec;

  // Map each FFT bin to a frequency, then to a band index (or -1 if outside).
  const nyquist = sampleRate / 2;
  const binToBand = new Int32Array(nFreqs);
  for (let k = 0; k < nFreqs; k++) {
    const freq = (k * sampleRate) / nFft;
    if (freq < BAND_EDGES_HZ[0]! || freq >= BAND_EDGES_HZ[NUM_BANDS]!) {
      binToBand[k] = -1;
      continue;
    }
    // Also clamp to nyquist — bins above nyquist should not exist (k <= nFft/2),
    // but this guards against pathological configs.
    if (freq > nyquist) {
      binToBand[k] = -1;
      continue;
    }
    let band = -1;
    for (let b = 0; b < NUM_BANDS; b++) {
      if (freq >= BAND_EDGES_HZ[b]! && freq < BAND_EDGES_HZ[b + 1]!) {
        band = b;
        break;
      }
    }
    if (band < 0) {
      binToBand[k] = -1;
    } else if (band > NUM_BANDS - 1) {
      binToBand[k] = NUM_BANDS - 1;
    } else {
      binToBand[k] = band;
    }
  }

  const frames: PhonemeFingerprint[] = new Array(nFrames);
  for (let f = 0; f < nFrames; f++) {
    const bands = new Float32Array(NUM_BANDS);
    const rowOffset = f * nFreqs;
    for (let k = 0; k < nFreqs; k++) {
      const b = binToBand[k]!;
      if (b < 0) continue;
      const m = magnitude[rowOffset + k]!;
      bands[b] = bands[b]! + m * m;
    }

    let totalEnergy = 0;
    let dominantBand = 0;
    let maxVal = -Infinity;
    for (let b = 0; b < NUM_BANDS; b++) {
      const v = bands[b]!;
      totalEnergy += v;
      if (v > maxVal) {
        maxVal = v;
        dominantBand = b;
      }
    }

    let category: PhonemeCategory;
    if (totalEnergy < silenceThreshold) {
      category = "silence";
    } else if (dominantBand >= 1 && dominantBand <= 4) {
      category = "voiced";
    } else if (dominantBand >= 5 && dominantBand <= 7) {
      category = "fricative";
    } else {
      category = "mixed";
    }

    frames[f] = { bands, totalEnergy, category, dominantBand };
  }

  return {
    frames,
    sampleRate,
    nFft,
    hopLength,
  };
}
