// @a0n/gnosis/distributed-inference-host/src/pipeline_pneuma.ts
/**
 * PneumaPipeline — orchestrates Whisper STT inference across N layer
 * stations on the gnosis mesh.
 *
 * Mirrors {@link Pipeline.pipelinedPrefill} from `./pipeline.ts` (the LLM
 * coordinator) but specialised for the encoder/decoder shape Whisper has:
 *
 *   - The **encoder** runs ONCE per utterance. The mel spectrogram is
 *     embedded on the entry station, the residual stream walks through the
 *     intermediate layer-range stations one hop at a time, and the exit
 *     station applies the final layer norm. There is no autoregressive
 *     loop here; the cost model is single-pass over a fixed-length input
 *     (`n_audio_ctx = 1500` frames).
 *
 *   - The **decoder** is autoregressive. Per generated token, the residual
 *     would walk through every decoder layer-range station once. Whisper's
 *     cross-attention K/V is computed once per utterance from the cached
 *     encoder output and reused for every step, so we ship the encoder
 *     tensor to each decoder station via `loadAudioContext` BEFORE the
 *     first `decodeStep`.
 *
 * Shape contracts (mirrors `WhisperEncoder` / `WhisperDecoder` in the
 * Rust crate):
 *
 *   mel              — Float32Array of length `n_mels * n_frames`
 *                       (typically `80 * 3000 = 240_000`).
 *   encoder residual — `(n_audio_ctx, encoder_hidden_dim)` row-major;
 *                       length = `n_audio_ctx * encoder_hidden_dim`
 *                       (typically `1500 * 384 = 576_000` for tiny).
 *   logits           — Float32Array of length `vocab_size` per decode step.
 *
 * Wire contract (per the brief; the actual CF Worker is Channel ζ):
 *
 *   POST /embed              — entry encoder station: mel → residual
 *   POST /forward-range      — middle encoder stations: residual → residual
 *   POST /finalize           — exit encoder station: residual → encoder out
 *   POST /load-audio-context — every decoder station, once per utterance
 *   POST /decode-step        — every decoder station, every token
 *
 * Endpoint surface lives on {@link Station} alongside the LLM endpoints
 * (`forward`, `lmHead`, …). New methods on Station are additive — the LLM
 * pipeline is untouched.
 *
 * v1 scope: encoder sharding + greedy decoding via a single decoder
 * station. Multi-station decoder sharding is stubbed below; each
 * autoregressive step over N decoder stations follows exactly the
 * per-token loop in {@link Pipeline.generate}, but the v1 demo only
 * needs the encoder spread across stations (the costly part) — the
 * decoder for whisper-tiny is small enough to fit on a single Worker.
 *
 * See: `PNEUMA_PHASE_2.md` § Channel δ.
 */

import type { Station } from './station';
import { MemoClient, type MemoFetcher } from './memo-client';

// ── Public types ──────────────────────────────────────────────────────────

/**
 * ER=EPR teleportation cache opt-in. When present at construction, the
 * pipeline fires `MemoClient.configure({enabled: true, ...})` against every
 * station that exposes an HTTP-shaped endpoint. Stations without /memo/*
 * support (404) are silently skipped.
 *
 * Bench-validated production defaults from `bin/bench-teleportation.rs`:
 *   - tauSquared: 1e-4   (cosine ≥ 0.999 vs baseline at the bench knot)
 *   - maxEntries: 8192   (matches the bench harness; ~8 MiB per station at hidden=3584)
 *
 * IMPORTANT (per the agent caveat): the cache only helps SPLIT-path callers
 * (`forward_attn_split` + `forward_ffn_split` → `cached_mat_vec`). Stations
 * served by the unchunked single-call `forward_range_chunk` path (e.g.
 * fat-station's `/forward`) do not route through the cache and will see
 * zero benefit even with `enableTeleportation` set. Wire this against the
 * wasm CF Worker stations.
 */
export interface TeleportationConfig {
  tauSquared?: number;
  maxEntries?: number;
}

/**
 * Duck-typed surface for stations that expose a fetch endpoint + base URL.
 * `Station` (HTTP) satisfies this structurally; `StationUds` and the flow
 * variants do not (UDS has no HTTP base URL). The teleportation boot path
 * uses this guard to decide which stations get a `/memo/configure` call.
 */
interface StationWithHttpFetcher {
  endpoint: MemoFetcher;
  baseUrl: string;
}

function asHttpFetcherStation(station: unknown): StationWithHttpFetcher | null {
  if (!station || typeof station !== 'object') return null;
  const s = station as { endpoint?: unknown; baseUrl?: unknown };
  if (
    s.endpoint &&
    typeof s.endpoint === 'object' &&
    typeof (s.endpoint as { fetch?: unknown }).fetch === 'function' &&
    typeof s.baseUrl === 'string'
  ) {
    return { endpoint: s.endpoint as MemoFetcher, baseUrl: s.baseUrl };
  }
  return null;
}

/** Tunable knobs for {@link PneumaPipeline}. */
export interface PneumaPipelineConfig {
  /**
   * Encoder layer-range stations, ordered. The first station is the entry
   * (owns the conv stem + pos-embed table), the last station is the exit
   * (owns `ln_final.{weight,bias}`). For a single-station deploy, pass
   * one station that owns the full layer range.
   */
  encoderStations: Station[];
  /**
   * Decoder layer-range stations, ordered. Same convention: first station
   * owns token-embed + pos-embed, last station owns the lm-head (tied to
   * the token-embed table for Whisper).
   *
   * v1 supports a single decoder station fully. Multi-station decoder
   * sharding is implemented as a per-token forward-range walk, mirroring
   * {@link Pipeline.generate}'s autoregressive loop — see the TODO in
   * {@link transcribe}.
   */
  decoderStations: Station[];
  /**
   * BOS token id used to seed greedy decoding.
   * Defaults to `50258` (multilingual whisper checkpoint). Use `50257`
   * for the `*.en` checkpoints, or any other special-token id for
   * forced-prompt experiments. Mirrors `WasmWhisperPipeline::set_bos_token`.
   */
  bosToken?: number;
  /**
   * Token id that ends generation. Defaults to `50257` (`<|endoftext|>`
   * for whisper). The transcribe loop stops as soon as this id is sampled.
   */
  eosToken?: number;
  /** Optional request id for tracing across station calls. */
  requestId?: string;
  /**
   * Optional sampler. If unspecified, argmax sampling is used. Receives
   * the logits Float32Array for the most recent token and returns the
   * chosen token id.
   */
  sample?: (logits: Float32Array) => number;
  /**
   * OFF by default. When set, the pipeline enables the ER=EPR
   * teleportation cache on every station that supports `/memo/configure`
   * during construction. Stations that 404 or otherwise reject the call
   * are silently skipped. See {@link TeleportationConfig} for caveats —
   * the cache only helps SPLIT-path stations (wasm Workers), not
   * fat-station's unchunked `/forward`.
   *
   * Bench-validated knobs (omit to inherit server defaults):
   *   { tauSquared: 1e-4, maxEntries: 8192 }
   */
  enableTeleportation?: TeleportationConfig;
}

/**
 * Per-station outcome of the teleportation boot fan-out. Surfaced by
 * {@link PneumaPipeline.teleportationReady} so callers (and tests) can
 * observe which stations were configured, which were skipped (no HTTP
 * endpoint), and which 404'd or failed.
 */
export interface TeleportationBootOutcome {
  stationIndex: number;
  kind: 'encoder' | 'decoder';
  /** "configured" — server accepted; "skipped" — no HTTP endpoint or no /memo route; "error" — call threw. */
  status: 'configured' | 'skipped' | 'error';
  error?: string;
}

/** Greedy `argmax` over a logits slice. Default sampler. */
function argmax(logits: Float32Array): number {
  let best = 0;
  let bestVal = -Infinity;
  for (let i = 0; i < logits.length; i++) {
    const v = logits[i]!;
    if (v > bestVal) {
      bestVal = v;
      best = i;
    }
  }
  return best;
}

/** True iff `arr` contains any NaN — early-abort signal mirroring `pipeline.ts`. */
function hasNaN(arr: Float32Array): boolean {
  for (let i = 0; i < arr.length; i++) {
    if (Number.isNaN(arr[i]!)) return true;
  }
  return false;
}

// ── PneumaPipeline ────────────────────────────────────────────────────────

/**
 * Mesh-sharded Whisper STT coordinator.
 *
 * Construction is cheap — no station calls happen until {@link encodeAudio}
 * or {@link transcribe}. The pipeline is stateless across utterances; the
 * decoder stations own their own KV caches, cleared by `loadAudioContext`.
 */
export class PneumaPipeline {
  private readonly encoderStations: Station[];
  private readonly decoderStations: Station[];
  private readonly bosToken: number;
  private readonly eosToken: number;
  private readonly requestId: string;
  private readonly sample: (logits: Float32Array) => number;
  /**
   * Promise resolves to the per-station outcome list when
   * `enableTeleportation` was set; null when the cache opt-in was omitted.
   * Construction kicks off the fan-out fire-and-forget — callers that need
   * to await readiness use {@link teleportationReady}.
   */
  private readonly teleportationBoot: Promise<TeleportationBootOutcome[]> | null;

  /**
   * @param encoderStations — each owns a contiguous encoder layer range.
   * @param decoderStations — each owns a contiguous decoder layer range.
   *
   * Both lists must be non-empty. The two-arg form mirrors the brief's
   * concise constructor signature; pass a {@link PneumaPipelineConfig}
   * via the static {@link PneumaPipeline.fromConfig} factory for full
   * tunability.
   */
  constructor(encoderStations: Station[], decoderStations: Station[]);
  constructor(config: PneumaPipelineConfig);
  constructor(
    arg1: Station[] | PneumaPipelineConfig,
    arg2?: Station[],
  ) {
    let cfg: PneumaPipelineConfig;
    if (Array.isArray(arg1)) {
      if (!arg2 || !Array.isArray(arg2)) {
        throw new Error(
          'PneumaPipeline: positional constructor requires (encoderStations, decoderStations)',
        );
      }
      cfg = { encoderStations: arg1, decoderStations: arg2 };
    } else {
      cfg = arg1;
    }
    if (!cfg.encoderStations.length) {
      throw new Error('PneumaPipeline: at least one encoder station is required');
    }
    if (!cfg.decoderStations.length) {
      throw new Error('PneumaPipeline: at least one decoder station is required');
    }
    this.encoderStations = cfg.encoderStations.slice();
    this.decoderStations = cfg.decoderStations.slice();
    this.bosToken = cfg.bosToken ?? 50258;
    this.eosToken = cfg.eosToken ?? 50257;
    this.requestId = cfg.requestId ?? `pneuma-${Date.now()}`;
    this.sample = cfg.sample ?? argmax;

    // Teleportation cache opt-in. OFF by default — only fired when the
    // caller explicitly sets `enableTeleportation`. Stations without an
    // HTTP-shaped endpoint (UDS / flow / mocks lacking endpoint+baseUrl)
    // are skipped without contacting the wire. 404s and other server
    // errors are swallowed so a station that doesn't ship the /memo/* route
    // does not poison the pipeline boot.
    this.teleportationBoot = cfg.enableTeleportation
      ? this.bootTeleportation(cfg.enableTeleportation)
      : null;
  }

  /**
   * Fan out `MemoClient.configure({ enabled: true, ... })` against every
   * station in {encoder ∪ decoder} that exposes an HTTP fetcher. Failures
   * are isolated per-station and recorded in the outcome list — the
   * promise itself resolves on success of the fan-out (it never rejects).
   *
   * Hot-path latency: configure is one POST per station, fired in parallel.
   * Boot-time only; not on the per-token path.
   */
  private async bootTeleportation(
    cfg: TeleportationConfig,
  ): Promise<TeleportationBootOutcome[]> {
    const outcomes: TeleportationBootOutcome[] = [];
    const tasks: Promise<void>[] = [];

    const fire = (
      station: Station,
      stationIndex: number,
      kind: 'encoder' | 'decoder',
    ): void => {
      const http = asHttpFetcherStation(station);
      if (!http) {
        outcomes.push({ stationIndex, kind, status: 'skipped' });
        return;
      }
      const memo = new MemoClient({ endpoint: http.endpoint, baseUrl: http.baseUrl });
      const body: { enabled: boolean; tauSquared?: number; maxEntries?: number } = {
        enabled: true,
      };
      if (typeof cfg.tauSquared === 'number') body.tauSquared = cfg.tauSquared;
      if (typeof cfg.maxEntries === 'number') body.maxEntries = cfg.maxEntries;
      tasks.push(
        memo.configure(body).then(
          () => {
            outcomes.push({ stationIndex, kind, status: 'configured' });
          },
          (err: unknown) => {
            const msg = err instanceof Error ? err.message : String(err);
            // 404 = station doesn't ship /memo/configure; skip silently in
            // the failure-mode bucket as 'skipped' so test/console output
            // doesn't read "error" for an expected-no-route station.
            if (msg.includes(' 404')) {
              outcomes.push({ stationIndex, kind, status: 'skipped' });
            } else {
              outcomes.push({ stationIndex, kind, status: 'error', error: msg });
            }
          },
        ),
      );
    };

    this.encoderStations.forEach((s, i) => fire(s, i, 'encoder'));
    this.decoderStations.forEach((s, i) => fire(s, i, 'decoder'));

    await Promise.all(tasks);
    return outcomes;
  }

  /**
   * Awaitable handle on the teleportation boot fan-out. Resolves to the
   * per-station outcome list, or null when `enableTeleportation` was not
   * passed to the constructor. Tests use this to assert that configure
   * was (or was not) called per station.
   */
  teleportationReady(): Promise<TeleportationBootOutcome[] | null> {
    return this.teleportationBoot ?? Promise.resolve(null);
  }

  /** Static factory mirroring {@link Pipeline}'s config-object constructor. */
  static fromConfig(config: PneumaPipelineConfig): PneumaPipeline {
    return new PneumaPipeline(config);
  }

  /** First (entry) encoder station — runs conv stem + pos embed. */
  get encoderEntry(): Station {
    return this.encoderStations[0]!;
  }

  /** Last (exit) encoder station — runs the final layer norm. */
  get encoderExit(): Station {
    return this.encoderStations[this.encoderStations.length - 1]!;
  }

  /** First (entry) decoder station — owns the token-embed table. */
  get decoderEntry(): Station {
    return this.decoderStations[0]!;
  }

  /** Last (exit) decoder station — owns the lm-head + final norm. */
  get decoderExit(): Station {
    return this.decoderStations[this.decoderStations.length - 1]!;
  }

  /**
   * Run the encoder by fanning out layer ranges to stations.
   *
   * `audio` is a 16 kHz mono Float32Array — but note that Whisper consumes
   * a log-mel spectrogram, not raw PCM. Hosts that already have raw audio
   * MUST run `extractMelSpectrogram` (from `./mel-spectrogram.ts`) before
   * calling this method, OR pass the mel directly via {@link encodeMel}.
   *
   * The `audio` parameter is currently treated as a pre-extracted mel
   * spectrogram for v1 (length must equal `n_mels * n_frames`). The
   * "raw audio in, mel inside" path is deferred to Channel ε (the
   * pneuma-duplex CLI) — host-side mel extraction matches HF Python within
   * 1e-3 already, so the cost-bearing decision was to keep the mel-spec
   * extractor on the host (no FFT in WASM) and let `encodeAudio` accept
   * either form.
   *
   * Returns the post-norm encoder output `(n_audio_ctx, encoder_hidden_dim)`
   * as a Float32Array. The host typically caches this and reuses it for
   * many `transcribe` calls (e.g. forced-prompt or beam variants).
   *
   * Step-by-step, mirroring {@link Pipeline.pipelinedPrefill}:
   *
   *   1. POST /embed              to encoderStations[0]   (mel → residual)
   *   2. for k in 1..N-1:
   *        POST /forward-range    to encoderStations[k]   (residual → residual)
   *   3. POST /finalize           to encoderStations[N-1] (residual → output)
   *
   * Single-station case (N == 1): the entry == exit, so we run /embed
   * then /finalize on the same station; /forward-range is unused. The
   * station's full layer range is still walked inside the embed handler
   * by convention (the Worker's /embed implementation is responsible for
   * running both the conv stem AND the encoder blocks when there's only
   * one station — it's an "all-in-one" form). We surface a
   * `forward-range` call only when N > 1 so the single-station path
   * stays a single round trip.
   */
  async encodeAudio(audio: Float32Array): Promise<Float32Array> {
    if (!audio.length) {
      throw new Error('PneumaPipeline.encodeAudio: audio buffer must be non-empty');
    }

    // Stage 1: embed on the entry station. Returns the post-conv residual
    // ready to feed the encoder blocks.
    let residual = await this.encoderEntry.embedAudio({
      mel: audio,
      requestId: this.requestId,
    });

    if (hasNaN(residual)) {
      throw new Error('PneumaPipeline.encodeAudio: NaN poison out of /embed');
    }

    // Stage 2: walk middle stations one hop at a time. The encoder is
    // stateless across hops (no KV cache) so this is a straight
    // residual-passing chain — no Promise.race, no in-flight bookkeeping,
    // because the same residual is the input to the next station.
    //
    // Mirrors the autoregressive single-token loop at the tail of
    // {@link Pipeline.generate}: per stage, await the call, NaN-guard,
    // forward to the next stage. The encoder is the prefill phase of an
    // utterance; the decoder is the per-token generate phase below.
    for (let k = 1; k < this.encoderStations.length - 1; k++) {
      residual = await this.encoderStations[k]!.forwardEncoderRange({
        residual,
        requestId: this.requestId,
      });
      if (hasNaN(residual)) {
        throw new Error(
          `PneumaPipeline.encodeAudio: NaN poison out of /forward-range on station ${k}`,
        );
      }
    }

    // Stage 3: finalize on the exit station. For N >= 2, the exit station
    // ALSO runs its own layer range first (the /finalize handler should
    // run forward_layer_range over its pinned range, then ln_final). For
    // N == 1, /embed already covered the layer range and /finalize is the
    // only remaining work; same handler shape.
    //
    // The entry == exit special case (N == 1) skips the inter-station
    // residual hop entirely.
    if (this.encoderStations.length > 1) {
      // Hand the residual coming out of the last middle station (or
      // straight from /embed when N == 2) to the exit station's
      // /forward-range first, then /finalize.
      residual = await this.encoderExit.forwardEncoderRange({
        residual,
        requestId: this.requestId,
      });
      if (hasNaN(residual)) {
        throw new Error(
          'PneumaPipeline.encodeAudio: NaN poison out of /forward-range on exit station',
        );
      }
    }

    const output = await this.encoderExit.finalizeEncoder({
      residual,
      requestId: this.requestId,
    });
    if (hasNaN(output)) {
      throw new Error('PneumaPipeline.encodeAudio: NaN poison out of /finalize');
    }
    return output;
  }

  /**
   * Run greedy autoregressive decoding using the cached encoder output.
   *
   * `encoderOutput` is the row-major `(n_audio_ctx, encoder_hidden_dim)`
   * tensor returned by {@link encodeAudio}. This method:
   *
   *   1. Calls `loadAudioContext` on every decoder station once. The
   *      decoder station handler is responsible for invoking
   *      `WhisperDecoder::prepare_audio_context` to populate its frozen
   *      cross-attention K/V cache and to clear its self-attention KV
   *      cache (per the contract on {@link Station.loadAudioContext}).
   *
   *   2. Seeds with `bosToken` and runs up to `maxTokens` decode steps:
   *
   *        for step in 0..maxTokens:
   *          for each decoder station k:
   *            logits = await station[k].decodeStep({ token, pos: step })
   *          next = sample(logits)
   *          if next == eosToken: stop
   *          emit next; token = next
   *
   *      For v1 we run a single `decodeStep` call per token against the
   *      decoder ENTRY station (which is identically the exit when there's
   *      only one decoder station). Multi-station decoder sharding requires
   *      the per-station decoder handler to expose `forwardDecoderRange`
   *      + a separate `lmHead` step — see the TODO below. The v1 contract
   *      accepts that decoder sharding is a follow-on; whisper-tiny's
   *      decoder is 4 layers and fits comfortably on one Worker.
   *
   * Returns the generated token ids (BOS not included). On `maxTokens
   * == 0` returns an empty array. `eosToken` is also excluded from the
   * output, matching `WasmWhisperPipeline::generate`.
   */
  async transcribe(
    encoderOutput: Float32Array,
    maxTokens: number = 256,
  ): Promise<number[]> {
    if (!Number.isInteger(maxTokens) || maxTokens < 0) {
      throw new Error(
        `PneumaPipeline.transcribe: maxTokens must be a non-negative integer (got ${String(maxTokens)})`,
      );
    }
    if (!encoderOutput.length) {
      throw new Error('PneumaPipeline.transcribe: encoderOutput must be non-empty');
    }
    if (maxTokens === 0) {
      return [];
    }

    // Step 1: prime cross-attn + reset self-attn on every decoder station.
    // We fire all calls in parallel — the stations are independent.
    // Mirror of {@link Pipeline.pipelinedPrefill}'s cobordism fan-out.
    await Promise.all(
      this.decoderStations.map((s) =>
        s.loadAudioContext({
          encoderOutput,
          requestId: this.requestId,
        }),
      ),
    );

    // Step 2: greedy decoding loop. Each step is a per-token round trip.
    //
    // TODO(channel-δ): multi-station decoder sharding. The current loop
    // runs one /decode-step call against the decoder entry station, which
    // assumes that station owns the full decoder pipeline (token embed,
    // all layers, lm-head). For mesh-sharded decoders, replace the inner
    // body with a per-station forward-range walk + a final lm-head call
    // — exactly the shape of {@link Pipeline.generate}'s per-token loop.
    // Stub left in place; whisper-tiny.en has 4 decoder layers and fits
    // on one Worker, so v1 prioritises encoder sharding.
    const out: number[] = [];
    let prev = this.bosToken;
    const decoderStation = this.decoderEntry;

    for (let step = 0; step < maxTokens; step++) {
      const logits = await decoderStation.decodeStep({
        token: prev,
        pos: step,
        requestId: this.requestId,
      });
      if (hasNaN(logits)) {
        // NaN poison stops generation — emit what we have so far.
        return out;
      }
      const next = this.sample(logits);
      if (next === this.eosToken) {
        break;
      }
      out.push(next);
      prev = next;
    }

    return out;
  }

  /**
   * Convenience: encode + transcribe in one shot. Mirrors the monolithic
   * `WasmWhisperPipeline` flow (encode → generate) for hosts that don't
   * need to cache the encoder output across multiple decode passes.
   */
  async encodeAndTranscribe(
    audio: Float32Array,
    maxTokens: number = 256,
  ): Promise<number[]> {
    const encoderOutput = await this.encodeAudio(audio);
    return this.transcribe(encoderOutput, maxTokens);
  }

  /**
   * Streaming transcribe — emit running transcripts as audio arrives, so
   * the user (or a downstream LLM) can react before the speaker stops.
   *
   * `audioStream` is an async iterable of mel chunks; each chunk is a
   * row-major `(n_mels, chunk_n_frames)` Float32Array. Callers typically
   * produce these from a 16 kHz mono PCM source by extracting log-mel
   * frames every ~250-1000 ms (1 s ≈ 100 mel frames). The caller picks
   * the chunk cadence: smaller chunks → lower partial-emit latency at the
   * cost of more encoder recomputes (v1's encoder re-runs over the whole
   * 30 s mel ring per chunk; see `forward_streaming_chunk` in the Rust
   * crate).
   *
   * Per chunk:
   *   1. Send the chunk to the encoder entry station's `/encode-streaming`
   *      handler. For multi-station encoder deploys, only the entry
   *      station owns the streaming mel ring — the partial encoder output
   *      it returns already accounts for the full encoder pipeline (the
   *      handler runs the streaming finalize internally).
   *      TODO(channel-δ): multi-station streaming needs the entry station
   *      to ship the partial residual through `/forward-range` of every
   *      middle station and `/finalize` of the exit station, just like
   *      `encodeAudio`. v1 supports the single-station encoder path,
   *      which matches the v1 contract on `transcribe`.
   *   2. Re-prime cross-attention on every decoder station against the
   *      new partial encoder output.
   *   3. Run greedy decoding from BOS until the first `eosToken` (or
   *      `maxTokensPerPartial`), with `resetSelfKV: true` on the first
   *      decode step so we don't leak self-KV state from the previous
   *      partial pass. Subsequent steps within the same partial keep the
   *      cache (positions advance normally).
   *   4. Yield (and fire `onPartial` for) the running transcript.
   *
   * The generator's last yield is the final transcript — once the audio
   * stream is exhausted, no further yields occur. The return value of an
   * async generator is `undefined`; the per-iteration value is the
   * partial transcript.
   *
   * `onPartial` is a convenience callback fired with the same array right
   * before the yield, for callers that don't want to drive a `for await`
   * loop themselves (e.g. piping to a render function).
   *
   * NOTE: For latency reasons we do NOT fan the chunk through any encoder
   * middle stations from the host side; the entry station's
   * `/encode-streaming` handler is responsible for running the full
   * encoder and returning the partial output ready for the decoder. This
   * keeps the per-chunk round-trip count at `1 + decoderStations.length`,
   * matching the cost model of `encodeAudio` + `transcribe` for one full
   * pass.
   *
   * Constraints:
   *   - This method does not call `encodeAudio`; the caller-supplied chunk
   *     stream replaces it. To go back to one-shot transcription,
   *     use `transcribe` or `encodeAndTranscribe`.
   *   - `audioStream` may yield empty Float32Array chunks; they are
   *     treated as keep-alives and skipped (no encoder call, no yield).
   *
   * @param audioStream chunks of `(n_mels, chunk_frames)` mel features,
   *                    in utterance order.
   * @param onPartial   optional sync callback fired on each partial.
   * @param maxTokensPerPartial cap on decode steps per partial pass.
   *                            Defaults to 256, same as `transcribe`.
   */
  async *transcribeStreaming(
    audioStream: AsyncIterable<Float32Array>,
    onPartial?: (partial: number[]) => void,
    maxTokensPerPartial: number = 256,
  ): AsyncIterableIterator<number[]> {
    if (!Number.isInteger(maxTokensPerPartial) || maxTokensPerPartial < 0) {
      throw new Error(
        `PneumaPipeline.transcribeStreaming: maxTokensPerPartial must be a non-negative integer (got ${String(maxTokensPerPartial)})`,
      );
    }

    let chunkIndex = 0;
    let lastPartial: number[] = [];

    for await (const melChunk of audioStream) {
      if (!melChunk.length) {
        // Keep-alive. Not a chunk boundary — skip.
        continue;
      }

      // 1. Stream the chunk through the encoder entry station. The Worker
      //    owns the 30 s mel ring; on chunkIndex == 0 it MUST reset its
      //    streaming state (per the Station.encodeStreamingChunk contract).
      const partialEncoderOutput = await this.encoderEntry.encodeStreamingChunk({
        melChunk,
        chunkIndex,
        requestId: this.requestId,
      });
      if (hasNaN(partialEncoderOutput)) {
        // Poison kills the stream — yield whatever we last produced and
        // stop. This mirrors `encodeAudio`'s NaN-guard behaviour.
        return;
      }

      // 2. Re-prime cross-attention on every decoder station with the
      //    fresh partial encoder output. Stations are independent → fan
      //    out in parallel, mirroring `transcribe`.
      await Promise.all(
        this.decoderStations.map((s) =>
          s.loadAudioContext({
            encoderOutput: partialEncoderOutput,
            requestId: this.requestId,
          }),
        ),
      );

      // 3. Greedy decode against this partial. Self-KV cache is wiped on
      //    the FIRST step (resetSelfKV: true) so we don't leak the
      //    previous partial's growing cache; subsequent steps within
      //    this partial pass build the cache normally.
      const partial: number[] = [];
      let prev = this.bosToken;
      const decoderStation = this.decoderEntry;

      for (let step = 0; step < maxTokensPerPartial; step++) {
        const logits = await decoderStation.decodeStep({
          token: prev,
          pos: step,
          resetSelfKV: step === 0,
          requestId: this.requestId,
        });
        if (hasNaN(logits)) {
          // NaN aborts THIS partial; we still emit what we generated
          // before the poison so the caller sees a usable prefix.
          break;
        }
        const next = this.sample(logits);
        if (next === this.eosToken) {
          break;
        }
        partial.push(next);
        prev = next;
      }

      lastPartial = partial;
      if (onPartial) {
        try {
          onPartial(partial);
        } catch {
          // Subscriber bugs must not break the generator. Swallow and
          // keep going.
        }
      }
      yield partial;
      chunkIndex += 1;
    }

    // No more audio. The caller already received the last partial via
    // the loop's final yield, so we don't yield it twice — but we keep
    // the local around in case future versions add a final pass (e.g.
    // a temperature-bumped re-decode after the speaker stops).
    void lastPartial;
  }
}
