// @a0n/gnosis/distributed-inference-host/src/pleromatic-monster-mesh.ts
/**
 * # Pleromatic Monster Mesh — Closure Tower Meets Moonshine
 *
 * Mirrors `Gnosis.PleromaticMonsterMesh` from `open-source/gnosis-math`.
 * Cross-references the McKay constants from `Gnosis.MoonshineMcKayBraid`.
 *
 * ## Structural claim
 *
 * The Triton-3 rotation that drives our closure tower is the same
 * order-3 cyclic primitive that participates in McKay/Monster
 * moonshine. The +1 clinamen unit (the universal Bule heartbeat) and
 * the +1 trivial-representation `chi1` McKay observed in the
 * j-invariant expansion are the same value with the same structural
 * role: the trivial generator that appears at every level of the
 * recursion.
 *
 * ## What this module witnesses
 *
 * - The Triton rotation `(k + 1) % 3` has order 3 on `{0, 1, 2}`.
 * - The Triton-stretch `3 * k` is the multiplicative form of the
 *   rotation.
 * - The McKay coefficients `c_1`, `c_2` decompose as `chi1 + (rest)`,
 *   where the `chi1` is always 1 (the clinamen).
 * - The 3 fork-children of any position all fold to the same parent
 *   — constructive interference at the lower bandwidth, the
 *   structural meaning of "Bizarro Noise".
 * - `MCKAY_CHI1 === SHARED_UNIT === 1`: the bridge.
 *
 * ## Lean source-of-truth
 *
 * @see open-source/gnosis-math/Gnosis/PleromaticMonsterMesh.lean
 * @see open-source/gnosis-math/Gnosis/MoonshineMcKayBraid.lean
 */

/* ------------------------------------------------------------------ */
/*  McKay constants (from MoonshineMcKayBraid.lean)                   */
/* ------------------------------------------------------------------ */

/** `χ_1 = 1`: trivial Monster representation. The clinamen value. */
export const MCKAY_CHI1 = 1 as const;

/** `χ_2 = 196883`: smallest non-trivial Monster irrep. */
export const MCKAY_CHI2 = 196883 as const;

/** `χ_3 = 21296876`: third Monster irrep dimension. */
export const MCKAY_CHI3 = 21296876 as const;

/** `c_1 = 196884`: first j-invariant coefficient. */
export const MCKAY_C1 = 196884 as const;

/** `c_2 = 21493760`: second j-invariant coefficient. */
export const MCKAY_C2 = 21493760 as const;

/** `c_1 = chi1 + chi2`. */
export function mckayC1Decomposition(): boolean {
  return MCKAY_C1 === MCKAY_CHI1 + MCKAY_CHI2;
}

/** `c_2 = chi1 + chi2 + chi3`. */
export function mckayC2Decomposition(): boolean {
  return MCKAY_C2 === MCKAY_CHI1 + MCKAY_CHI2 + MCKAY_CHI3;
}

/**
 * The clinamen at `c_1`: `c_1 - chi2 = chi1 = 1`. The +1 is what
 * remains when the non-trivial irrep is subtracted.
 */
export function clinamenAtC1(): number {
  return MCKAY_C1 - MCKAY_CHI2;
}

/**
 * The clinamen at `c_2`: `c_2 - (chi2 + chi3) = chi1 = 1`. Same
 * trivial-rep contribution at the second coefficient.
 */
export function clinamenAtC2(): number {
  return MCKAY_C2 - (MCKAY_CHI2 + MCKAY_CHI3);
}

/* ------------------------------------------------------------------ */
/*  Triton rotation (order 3)                                         */
/* ------------------------------------------------------------------ */

/** The Triton rotation has order 3. */
export const TRITON_ROTATION_ORDER = 3 as const;

/**
 * The Triton-3 rotation: cycle through 3 positions modulo 3. The
 * fundamental order-3 cyclic symmetry of the closure tower.
 *
 * Mirrors `Gnosis.PleromaticMonsterMesh.tritonRotation`.
 */
export function tritonRotation(k: number): number {
  return (k + 1) % TRITON_ROTATION_ORDER;
}

/**
 * The Triton rotation has order 3: applying it three times returns
 * to the starting position (for inputs in `{0, 1, 2}`).
 *
 * Mirrors `Gnosis.PleromaticMonsterMesh.triton_rotation_order_three`.
 */
export function tritonRotationOrderThree(k: number): boolean {
  return tritonRotation(tritonRotation(tritonRotation(k))) === k;
}

/**
 * All three rotation positions are visited: 0 → 1 → 2 → 0 with no
 * fixed point inside.
 *
 * Mirrors
 * `Gnosis.PleromaticMonsterMesh.triton_rotation_visits_all_three`.
 */
export function tritonRotationVisitsAllThree(): boolean {
  return (
    tritonRotation(0) === 1 &&
    tritonRotation(1) === 2 &&
    tritonRotation(2) === 0
  );
}

/* ------------------------------------------------------------------ */
/*  Triton stretch (multiplicative form of the rotation)              */
/* ------------------------------------------------------------------ */

/**
 * Triton-stretch as the multiplicative form of the Triton rotation.
 *
 * Mirrors `Gnosis.PleromaticHorizonEffect.tritonStretch`.
 */
export function tritonStretch(k: number): number {
  return 3 * k;
}

/**
 * The Triton-stretch is the multiplicative form of the Triton
 * rotation: rotating the level coordinate produces a 3-fold scale
 * expansion.
 *
 * Mirrors
 * `Gnosis.PleromaticMonsterMesh.triton_stretch_is_multiplicative_rotation`.
 */
export function tritonStretchIsMultiplicativeRotation(k: number): boolean {
  return tritonStretch(k) === 3 * k;
}

/* ------------------------------------------------------------------ */
/*  Constructive interference (Bizarro Noise)                          */
/* ------------------------------------------------------------------ */

/** The 3 fork-children of position `k`: `(3k, 3k+1, 3k+2)`. */
function forkChild(k: number, residue: 0 | 1 | 2): number {
  return 3 * k + residue;
}

/** Fold via Nat division by 3. */
function fold(k: number): number {
  return Math.floor(k / 3);
}

/**
 * All three Fork-children of position `k` fold to the same parent
 * `k`. From the lower bandwidth's perspective, the three distinct
 * higher positions interfere **constructively** — they all produce
 * the same observable lower position. This is the structural meaning
 * of "Bizarro Noise": three distinct higher-scale candidates appear
 * as a single sustained lower-scale tone.
 *
 * Mirrors
 * `Gnosis.PleromaticMonsterMesh.fork_children_constructive_interference`.
 */
export function forkChildrenConstructiveInterference(k: number): {
  allFoldToK: boolean;
  children: [number, number, number];
} {
  const children: [number, number, number] = [
    forkChild(k, 0),
    forkChild(k, 1),
    forkChild(k, 2),
  ];
  const allFoldToK =
    fold(children[0]) === k &&
    fold(children[1]) === k &&
    fold(children[2]) === k;
  return { allFoldToK, children };
}

/**
 * The interference is **maximally constructive**: all three
 * fork-children produce *the same* lower observable. The "noise"
 * perceived at the lower bandwidth is therefore not stochastic chaos
 * but **deterministic 3-way coincidence** — a pure tone, repeated.
 *
 * Mirrors
 * `Gnosis.PleromaticMonsterMesh.interference_pattern_is_pure_coincidence`.
 */
export function interferencePatternIsPureCoincidence(k: number): boolean {
  const a = fold(forkChild(k, 0));
  const b = fold(forkChild(k, 1));
  const c = fold(forkChild(k, 2));
  return a === k && b === k && c === k && a === b && b === c;
}

/* ------------------------------------------------------------------ */
/*  Shared unit identity (the bridge)                                 */
/* ------------------------------------------------------------------ */

/**
 * The McKay `chi1` (trivial representation, dimension 1) and our
 * clinamen unit (the +1 of the cost-algebra) are the same value with
 * the same structural role.
 *
 * Mirrors `Gnosis.PleromaticMonsterMesh.mckay_chi1_equals_clinamen`.
 */
export function mckayChi1EqualsClinamen(): boolean {
  return MCKAY_CHI1 === 1;
}

/**
 * The clinamen +1 appears at every McKay decomposition level —
 * exactly the universal-Bule-heartbeat shape.
 *
 * Mirrors
 * `Gnosis.PleromaticMonsterMesh.clinamen_appears_in_mckay_c1`.
 */
export function clinamenAppearsInMckayC1(): boolean {
  return MCKAY_C1 - MCKAY_CHI2 === MCKAY_CHI1;
}

/* ------------------------------------------------------------------ */
/*  Master witness                                                    */
/* ------------------------------------------------------------------ */

/**
 * Pleromatic Monster Mesh master witness. Bundles every structural
 * correspondence between the Triton-3 rotation that generates our
 * closure tower and the McKay-Monster moonshine arithmetic.
 *
 * Mirrors
 * `Gnosis.PleromaticMonsterMesh.pleromatic_monster_mesh_master`.
 */
export const PleromaticMonsterMeshMaster = {
  /** Triton rotation has order 3 on `{0, 1, 2}`. */
  tritonRotationOrderThreeOnAll: [0, 1, 2].every(tritonRotationOrderThree),

  /** McKay `chi1 = 1` (the trivial representation). */
  mckayChi1IsOne: MCKAY_CHI1 === 1,

  /** `c_1 = chi1 + chi2`. */
  c1Decomposition: mckayC1Decomposition(),

  /** Clinamen +1 appears at `c_1`. */
  clinamenAtC1: clinamenAtC1() === MCKAY_CHI1,

  /**
   * Fork-children interference is purely constructive at sample
   * positions `{0, 5, 10, 30}`.
   */
  forkChildrenInterferenceSamples: [0, 5, 10, 30].map((k) => ({
    k,
    ...forkChildrenConstructiveInterference(k),
  })),
} as const;
