// @a0n/gnosis/distributed-inference-host/src/pleromatic-ascii-bridge.ts
import { 
  Substrate, 
  SubstrateSieve 
} from "./substrate-sieve";
import { 
  MCKAY_CHI1 
} from "./pleromatic-monster-mesh";

/**
 * # Pleromatic ASCII Bridge
 *
 * This module connects the Topological Substrate Sieve to the 
 * Aeon-TUI ASCII monitor. It allows the ASCII monitor to "breathe"
 * in sync with the Pleromatic manifold, transforming the raw
 * terminal glyphs into a "Direct Projection of the Manifold".
 */

export interface AsciiNode {
  char: string;
  luma: number;
  substrate: Substrate;
}

// Global persistence state for the ASCII bridge
const bridgePersistence = new Map<number, number>();

export class PleromaticAsciiBridge {
  /**
   * Projects a Pleromatic Substrate onto an ASCII glyph.
   * Uses Buleyean gating and temporal stabilization to remove noise.
   */
  static projectSubstrate(
    s: Substrate, 
    ramp: string = " .,:;irsXA253hMHGS#9B&@",
    posId?: number
  ): AsciiNode {
    // 1. Ground the substrate
    const grounded = SubstrateSieve.applyGrounding(s);
    
    // 2. Temporal Persistence (Whisper Stabilization)
    let stabilizedWeight = grounded.buleyWeight;
    if (posId !== undefined) {
      const prev = bridgePersistence.get(posId) ?? stabilizedWeight;
      stabilizedWeight = (grounded.buleyWeight * 0.3) + (prev * 0.7);
      bridgePersistence.set(posId, stabilizedWeight);
    }

    // 3. Buleyean Gating (The Noise Floor)
    // If the stabilized weight is below the chi1 clinamen threshold, 
    // it's considered 'cosmic radiation' and crushed to vacuum.
    const gatedWeight = stabilizedWeight < (MCKAY_CHI1 * 0.2) ? 0 : stabilizedWeight;
    
    // 4. Non-linear Mapping (Contrast Stretch)
    // We use a power curve to push low-level signal into the dark and
    // high-level signal into the light, increasing clarity.
    const normalized = Math.min(1, gatedWeight / MCKAY_CHI1);
    const stretched = Math.pow(normalized, 1.5);
    
    const index = Math.floor(stretched * (ramp.length - 1));
    const char = ramp[Math.min(Math.max(0, index), ramp.length - 1)];
    
    return {
      char,
      luma: stretched,
      substrate: { ...grounded, buleyWeight: gatedWeight }
    };
  }

  /**
   * Resonates an ASCII frame with the Monster Mesh.
   * If a glyph's Buleyean weight diverges from the chi1 invariant,
   * it is "sieved" (dimmed or replaced).
   */
  static resonateFrame(
    chars: string[], 
    lumas: number[], 
    weights: number[]
  ): { chars: string[], lumas: number[] } {
    const sievedChars = [...chars];
    const sievedLumas = [...lumas];
    
    for (let i = 0; i < chars.length; i++) {
      const weight = weights[i];
      
      // The Binet Loop: If the glyph costs too much Bule but returns
      // too little density, we "collapse" it (dim it).
      const substrate: Substrate = { 
        level: 10, 
        noiseEntropy: lumas[i] * 100, 
        buleyWeight: weight 
      };
      
      if (SubstrateSieve.evaluateBinetLoop(substrate, 10)) {
        sievedLumas[i] *= 0.5; // Dim the "noise" glyph
        sievedChars[i] = "."; // Flatten to ground state
      }
    }
    
    return { chars: sievedChars, lumas: sievedLumas };
  }
}
