// @a0n/gnosis/distributed-inference-host/src/prefill-window.ts
import { randomUUID } from 'node:crypto';
import type {
  Pipeline,
  PipelinePrefillProgress,
  PipelineTokenEvent,
  PrefillResult,
} from './pipeline';

export type PrefillWindowState =
  | 'open'
  | 'warming'
  | 'ready'
  | 'consumed'
  | 'expired'
  | 'mismatch'
  | 'aborted';

export type PrefillTelemetryStatus =
  | 'hit'
  | 'miss'
  | 'mismatch'
  | 'expired'
  | 'disabled';

export interface PrefillMessage {
  role: string;
  content: string;
}

export interface PrefillWindowSnapshot {
  id: string;
  model: string;
  state: PrefillWindowState;
  warmed_token_count: number;
  expires_at: number;
  last_miss_reason: string | null;
  warming_ms?: number;
  kv_saved?: boolean;
}

export interface PrefillGenerationResult {
  tokens: Uint32Array;
  telemetry?: PrefillTelemetry;
}

export interface PrefillTelemetry {
  status: PrefillTelemetryStatus;
  tokens?: number;
  savedMs?: number;
  missReason?: string;
}

interface PrefillWindowRecord {
  id: string;
  model: string;
  messages: PrefillMessage[];
  state: PrefillWindowState;
  createdAt: number;
  updatedAt: number;
  expiresAt: number;
  version: number;
  warmedTokenCount: number;
  prefixTokens: number[];
  prefill: PrefillResult | null;
  warmPromise: Promise<void> | null;
  warmingMs: number | null;
  kvSaved: boolean;
  lastMissReason: string | null;
}

export interface PrefillWindowManagerConfig {
  pipeline: Pipeline;
  tokenizeMessages: (model: string, messages: readonly PrefillMessage[]) => number[];
  enabled?: boolean;
  maxWindows?: number;
  maxWarmTokens?: number;
  defaultTtlMs?: number;
  now?: () => number;
}

export interface PrefillGenerateArgs {
  model: string;
  messages: PrefillMessage[];
  maxNewTokens: number;
  prefillWindowId?: string | null;
  suppressTerminalProsody?: boolean;
  onPrefillProgress?: (progress: PipelinePrefillProgress) => void;
  onToken?: (event: PipelineTokenEvent) => void;
  onTelemetry?: (telemetry: PrefillTelemetry) => void;
}

function normalizeMessages(messages: readonly PrefillMessage[]): PrefillMessage[] {
  return messages.map((message) => ({
    role: String(message.role),
    content: String(message.content ?? ''),
  }));
}

function tokenPrefixMatches(
  prefixTokens: readonly number[],
  tokens: readonly number[],
): boolean {
  if (prefixTokens.length > tokens.length) return false;
  for (let i = 0; i < prefixTokens.length; i++) {
    if (prefixTokens[i] !== tokens[i]) return false;
  }
  return true;
}

function clampPositiveInteger(value: unknown, fallback: number): number {
  return typeof value === 'number' && Number.isInteger(value) && value > 0
    ? value
    : fallback;
}

/**
 * Handles the gnosis prefill Telemetry Headers workflow.
 */
export function prefillTelemetryHeaders(
  telemetry: PrefillTelemetry | undefined,
): Record<string, string> {
  if (!telemetry) return {};
  return {
    'X-Moonshine-Prefill': telemetry.status,
    ...(telemetry.tokens !== undefined
      ? { 'X-Moonshine-Prefill-Tokens': String(telemetry.tokens) }
      : {}),
    ...(telemetry.savedMs !== undefined
      ? { 'X-Moonshine-Prefill-Saved-Ms': String(telemetry.savedMs) }
      : {}),
    ...(telemetry.missReason
      ? { 'X-Moonshine-Prefill-Miss-Reason': telemetry.missReason }
      : {}),
  };
}

export class PrefillWindowManager {
  private readonly pipeline: Pipeline;
  private readonly tokenizeMessages: PrefillWindowManagerConfig['tokenizeMessages'];
  private readonly enabled: boolean;
  private readonly maxWindows: number;
  private readonly maxWarmTokens: number;
  private readonly defaultTtlMs: number;
  private readonly now: () => number;
  private readonly windows = new Map<string, PrefillWindowRecord>();
  private exclusiveTail: Promise<void> = Promise.resolve();

  constructor(config: PrefillWindowManagerConfig) {
    this.pipeline = config.pipeline;
    this.tokenizeMessages = config.tokenizeMessages;
    this.enabled = config.enabled ?? true;
    this.maxWindows = clampPositiveInteger(config.maxWindows, 1);
    this.maxWarmTokens = clampPositiveInteger(config.maxWarmTokens, 512);
    this.defaultTtlMs = clampPositiveInteger(config.defaultTtlMs, 15_000);
    this.now = config.now ?? Date.now;
  }

  get isEnabled(): boolean {
    return this.enabled;
  }

  async runExclusive<T>(fn: () => Promise<T>): Promise<T> {
    let release!: () => void;
    const previous = this.exclusiveTail;
    this.exclusiveTail = new Promise<void>((resolve) => {
      release = resolve;
    });
    await previous;
    try {
      return await fn();
    } finally {
      release();
    }
  }

  create(args: {
    model: string;
    messages: PrefillMessage[];
    ttlMs?: number;
  }): PrefillWindowSnapshot {
    this.assertEnabled();
    this.evictExpired();
    this.evictToCapacity(1);

    const now = this.now();
    const id = randomUUID();
    const window: PrefillWindowRecord = {
      id,
      model: args.model,
      messages: normalizeMessages(args.messages),
      state: 'open',
      createdAt: now,
      updatedAt: now,
      expiresAt: now + clampPositiveInteger(args.ttlMs, this.defaultTtlMs),
      version: 0,
      warmedTokenCount: 0,
      prefixTokens: [],
      prefill: null,
      warmPromise: null,
      warmingMs: null,
      kvSaved: false,
      lastMissReason: null,
    };
    this.windows.set(id, window);
    this.scheduleWarm(window, null);
    return this.snapshot(window);
  }

  update(
    id: string,
    args: { messages: PrefillMessage[]; ttlMs?: number },
  ): PrefillWindowSnapshot | null {
    this.assertEnabled();
    const window = this.getLiveWindow(id);
    if (!window) return null;

    const previousReady =
      window.state === 'ready' && window.prefill
        ? {
            prefixTokens: window.prefixTokens.slice(),
            prefill: window.prefill,
          }
        : null;
    window.messages = normalizeMessages(args.messages);
    window.updatedAt = this.now();
    window.expiresAt =
      window.updatedAt + clampPositiveInteger(args.ttlMs, this.defaultTtlMs);
    window.version++;
    window.state = 'open';
    window.lastMissReason = null;
    this.scheduleWarm(window, previousReady);
    return this.snapshot(window);
  }

  get(id: string): PrefillWindowSnapshot | null {
    const window = this.getLiveWindow(id);
    return window ? this.snapshot(window) : null;
  }

  async delete(id: string): Promise<boolean> {
    const window = this.windows.get(id);
    if (!window) return false;
    window.state = 'aborted';
    this.windows.delete(id);
    await this.pipeline.deleteKvSession(id).catch(() => false);
    return true;
  }

  async generate(args: PrefillGenerateArgs): Promise<PrefillGenerationResult> {
    if (!this.enabled) {
      const telemetry = args.prefillWindowId
        ? { status: 'disabled' as const, missReason: 'prefill windows disabled' }
        : undefined;
      if (telemetry) args.onTelemetry?.(telemetry);
      return {
        tokens: await this.runExclusive(() => this.generateCold(args)),
        telemetry,
      };
    }

    const id = args.prefillWindowId?.trim();
    if (!id) {
      return {
        tokens: await this.runExclusive(() => this.generateCold(args)),
      };
    }

    const existingWindow = this.windows.get(id);
    if (existingWindow && this.isExpired(existingWindow)) {
      existingWindow.state = 'expired';
      existingWindow.lastMissReason = 'expired';
      this.windows.delete(id);
      void this.pipeline.deleteKvSession(id).catch(() => false);
      const telemetry = { status: 'expired' as const, missReason: 'expired window' };
      args.onTelemetry?.(telemetry);
      return {
        tokens: await this.runExclusive(() => this.generateCold(args)),
        telemetry,
      };
    }

    const window = this.getLiveWindow(id);
    if (!window) {
      const telemetry = { status: 'miss' as const, missReason: 'unknown window' };
      args.onTelemetry?.(telemetry);
      return {
        tokens: await this.runExclusive(() => this.generateCold(args)),
        telemetry,
      };
    }

    if (window.warmPromise) {
      await window.warmPromise.catch(() => undefined);
    }

    return await this.runExclusive(async () => {
      const live = this.getLiveWindow(id);
      if (!live) {
        const telemetry = { status: 'expired' as const, missReason: 'expired window' };
        args.onTelemetry?.(telemetry);
        return {
          tokens: await this.generateCold(args),
          telemetry,
        };
      }

      if (live.state !== 'ready' || !live.prefill) {
        live.lastMissReason = `window ${live.state}`;
        const telemetry = { status: 'miss' as const, missReason: live.lastMissReason };
        args.onTelemetry?.(telemetry);
        return {
          tokens: await this.generateCold(args),
          telemetry,
        };
      }

      if (live.model !== args.model) {
        live.state = 'mismatch';
        live.lastMissReason = 'model mismatch';
        const telemetry = { status: 'mismatch' as const, missReason: live.lastMissReason };
        args.onTelemetry?.(telemetry);
        return {
          tokens: await this.generateCold(args),
          telemetry,
        };
      }

      const finalTokens = this.tokenizeMessages(args.model, args.messages);
      if (!tokenPrefixMatches(live.prefixTokens, finalTokens)) {
        live.state = 'mismatch';
        live.lastMissReason = 'final prompt does not start with warmed token prefix';
        const telemetry = { status: 'mismatch' as const, missReason: live.lastMissReason };
        args.onTelemetry?.(telemetry);
        return {
          tokens: await this.generateCold(args),
          telemetry,
        };
      }

      const restored = await this.pipeline.loadKvSession(id).catch(() => null);
      if (!restored || restored.prefixLength !== live.prefixTokens.length) {
        live.lastMissReason = 'kv restore failed';
        const telemetry = { status: 'miss' as const, missReason: live.lastMissReason };
        args.onTelemetry?.(telemetry);
        return {
          tokens: await this.generateCold(args),
          telemetry,
        };
      }

      const telemetry = {
        status: 'hit' as const,
        tokens: live.prefixTokens.length,
        savedMs: live.warmingMs ?? undefined,
      };
      args.onTelemetry?.(telemetry);
      let prefill = live.prefill;
      if (finalTokens.length > live.prefixTokens.length) {
        const suffix = finalTokens.slice(live.prefixTokens.length);
        prefill = await this.pipeline.pipelinedPrefill({
          promptTokens: new Uint32Array(suffix),
          startPosOffset: live.prefixTokens.length,
          onProgress: args.onPrefillProgress,
        });
      }

      const tokens = await this.pipeline.generateFromPrefill({
        promptTokens: new Uint32Array(finalTokens),
        prefill,
        maxNewTokens: args.maxNewTokens,
        onToken: args.onToken,
      });
      live.state = 'consumed';
      this.windows.delete(id);
      void this.pipeline.deleteKvSession(id).catch(() => false);
      return {
        tokens,
        telemetry,
      };
    });
  }

  private assertEnabled(): void {
    if (!this.enabled) {
      throw new Error('prefill windows disabled');
    }
  }

  private async generateCold(args: PrefillGenerateArgs): Promise<Uint32Array> {
    const promptTokens = this.tokenizeMessages(args.model, args.messages);
    return await this.pipeline.generate({
      promptTokens: new Uint32Array(promptTokens),
      maxNewTokens: args.maxNewTokens,
      suppressTerminalProsody: args.suppressTerminalProsody,
      onPrefillProgress: args.onPrefillProgress,
      onToken: args.onToken,
    });
  }

  private scheduleWarm(
    window: PrefillWindowRecord,
    previousReady: { prefixTokens: number[]; prefill: PrefillResult } | null,
  ): void {
    const version = window.version;
    const tokenized = this.tokenizeMessages(window.model, window.messages).slice(
      0,
      this.maxWarmTokens,
    );
    window.state = 'warming';
    window.prefixTokens = tokenized;
    window.warmedTokenCount = 0;
    window.prefill = null;
    window.kvSaved = false;

    window.warmPromise = this.runExclusive(async () => {
      const live = this.windows.get(window.id);
      if (!live || live.version !== version || live.state === 'aborted') return;
      if (this.isExpired(live)) {
        live.state = 'expired';
        return;
      }

      const startedAt = this.now();
      try {
        let prefill: PrefillResult;
        if (
          previousReady &&
          previousReady.prefixTokens.length > 0 &&
          tokenPrefixMatches(previousReady.prefixTokens, tokenized) &&
          (await this.pipeline.loadKvSession(live.id).catch(() => null))
        ) {
          const suffix = tokenized.slice(previousReady.prefixTokens.length);
          prefill =
            suffix.length > 0
              ? await this.pipeline.pipelinedPrefill({
                  promptTokens: new Uint32Array(suffix),
                  startPosOffset: previousReady.prefixTokens.length,
                })
              : previousReady.prefill;
        } else {
          prefill = await this.pipeline.pipelinedPrefill({
            promptTokens: new Uint32Array(tokenized),
          });
        }

        const kvSaved = await this.pipeline
          .saveKvSession(live.id, tokenized.length)
          .catch(() => false);
        if (live.version !== version || this.windows.get(live.id) !== live) return;
        live.prefill = prefill;
        live.warmedTokenCount = tokenized.length;
        live.warmingMs = Math.max(0, this.now() - startedAt);
        live.kvSaved = kvSaved;
        live.state = kvSaved ? 'ready' : 'open';
        live.lastMissReason = kvSaved ? null : 'station kv sessions unavailable';
      } catch (error) {
        if (live.version !== version || this.windows.get(live.id) !== live) return;
        live.state = 'open';
        live.lastMissReason = error instanceof Error ? error.message : String(error);
      }
    });
  }

  private getLiveWindow(id: string): PrefillWindowRecord | null {
    const window = this.windows.get(id);
    if (!window) return null;
    if (this.isExpired(window)) {
      window.state = 'expired';
      window.lastMissReason = 'expired';
      this.windows.delete(id);
      void this.pipeline.deleteKvSession(id).catch(() => false);
      return null;
    }
    return window;
  }

  private isExpired(window: PrefillWindowRecord): boolean {
    return this.now() >= window.expiresAt;
  }

  private evictExpired(): void {
    for (const window of this.windows.values()) {
      if (this.isExpired(window)) {
        window.state = 'expired';
        this.windows.delete(window.id);
        void this.pipeline.deleteKvSession(window.id).catch(() => false);
      }
    }
  }

  private evictToCapacity(incoming: number): void {
    while (this.windows.size + incoming > this.maxWindows) {
      const oldest = [...this.windows.values()].sort(
        (a, b) => a.updatedAt - b.updatedAt,
      )[0];
      if (!oldest) return;
      oldest.state = 'aborted';
      this.windows.delete(oldest.id);
      void this.pipeline.deleteKvSession(oldest.id).catch(() => false);
    }
  }

  private snapshot(window: PrefillWindowRecord): PrefillWindowSnapshot {
    return {
      id: window.id,
      model: window.model,
      state: window.state,
      warmed_token_count: window.warmedTokenCount,
      expires_at: window.expiresAt,
      last_miss_reason: window.lastMissReason,
      ...(window.warmingMs !== null ? { warming_ms: window.warmingMs } : {}),
      kv_saved: window.kvSaved,
    };
  }
}
