// @a0n/gnosis/distributed-inference-host/src/public-cli.ts
// Public CLI for `affectively-deblur`.
//
// Detects input type by file extension and routes to the appropriate pipeline:
//   - Images (.png, .jpg, .jpeg, .webp)  -> deblurAnyImage + encodeRecoveredAsPNG
//   - Videos (.mp4, .mov, .webm, .avi, .mkv) -> deblurVideo (ffmpeg-driven)
//   - Audio (.wav, .flac)                -> not yet wired (programmatic API only)
//
// Uses node:util.parseArgs — no external dependencies.

import { parseArgs } from 'node:util';
import { writeFile } from 'node:fs/promises';
import { extname } from 'node:path';
import { deblurAnyImage, encodeRecoveredAsPNG, type DeblurAnyImageOptions } from './deblur-any-image';
import { deblurVideo } from './video-deblur';
import type { InvisibleWatermarkOptions } from './invisible-watermark';

export interface CliArgs {
  input: string;
  output: string;
  // Deblur options
  sigma?: number | 'auto';
  method?: 'wiener' | 'richardson-lucy';
  channelMode?: 'rgb' | 'luminance';
  noiseFloor?: number | 'auto';
  bilateral?: boolean;
  // Video options
  workers?: number;
  // Invisible watermark options
  watermark?: boolean;
  watermarkSeed?: string;
  watermarkStrength?: number;
  watermarkPayload?: string;
  // Misc
  verbose?: boolean;
}

const HELP_TEXT = `Usage: affectively-deblur <input> <output> [options]

Image input  (.png, .jpg, .jpeg, .webp)  → recovered image
Video input  (.mp4, .mov, .webm, .avi)   → recovered video
Audio input  (.wav, .flac)               → recovered audio (not yet wired)

Options:
  --sigma <auto|N>           Blur σ. Default 'auto' (φ-attractor selector).
  --method <wiener|richardson-lucy>   Deblur kernel. Default wiener.
  --channel-mode <rgb|luminance>      RGB recovers all 3 channels; luminance
                                       recovers Y only and rejoins original chroma.
  --noise-floor <auto|N>     Wiener regularization. Default 1e-3.
  --bilateral                Enable edge-preserving post-pass on Y.
  --workers <N>              Worker thread count (video only). 0 = inline.
  --watermark                Enable invisible watermark embedding.
  --watermark-seed <text>    Seed for deterministic watermark layout.
  --watermark-strength <N>   Strength in [0.001, 0.08]. Default 0.02.
  --watermark-payload <text> Payload tag hashed into watermark id.
  -v, --verbose              Print per-step diagnostics.
  -h, --help                 Show this help.

Examples:
  affectively-deblur photo.jpg out.png --sigma auto
  affectively-deblur clip.mp4 out.mp4 --workers 4 --bilateral
  affectively-deblur photo.jpg out.png --sigma 2.0 --method richardson-lucy
`;

const IMAGE_EXTS = new Set(['.png', '.jpg', '.jpeg', '.webp']);
const VIDEO_EXTS = new Set(['.mp4', '.mov', '.webm', '.avi', '.mkv']);
const AUDIO_EXTS = new Set(['.wav', '.flac']);

function parseSigma(v: string | undefined): number | 'auto' | undefined {
  if (v === undefined) return undefined;
  if (v === 'auto') return 'auto';
  const n = Number(v);
  if (!Number.isFinite(n) || n <= 0) {
    throw new Error(`Invalid --sigma: ${v} (expect 'auto' or a positive number)`);
  }
  return n;
}

function parseNoiseFloor(v: string | undefined): number | 'auto' | undefined {
  if (v === undefined) return undefined;
  if (v === 'auto') return 'auto';
  const n = Number(v);
  if (!Number.isFinite(n) || n <= 0) {
    throw new Error(`Invalid --noise-floor: ${v} (expect 'auto' or a positive number)`);
  }
  return n;
}

function parseInteger(v: string | undefined, label: string): number | undefined {
  if (v === undefined) return undefined;
  const n = Number(v);
  if (!Number.isInteger(n) || n < 0) {
    throw new Error(`Invalid --${label}: ${v} (expect non-negative integer)`);
  }
  return n;
}

function parseWatermarkStrength(v: string | undefined): number | undefined {
  if (v === undefined) return undefined;
  const n = Number(v);
  if (!Number.isFinite(n) || n <= 0) {
    throw new Error(`Invalid --watermark-strength: ${v} (expect positive number)`);
  }
  return n;
}

function parseMethod(v: string | undefined): 'wiener' | 'richardson-lucy' | undefined {
  if (v === undefined) return undefined;
  if (v !== 'wiener' && v !== 'richardson-lucy') {
    throw new Error(`Invalid --method: ${v} (expect 'wiener' or 'richardson-lucy')`);
  }
  return v;
}

function parseChannelMode(v: string | undefined): 'rgb' | 'luminance' | undefined {
  if (v === undefined) return undefined;
  if (v !== 'rgb' && v !== 'luminance') {
    throw new Error(`Invalid --channel-mode: ${v} (expect 'rgb' or 'luminance')`);
  }
  return v;
}

/**
 * Runs the gnosis command-line workflow.
 */
export async function main(argv: string[]): Promise<number> {
  let values: Record<string, string | boolean | undefined>;
  let positionals: string[];

  try {
    const parsed = parseArgs({
      args: argv,
      allowPositionals: true,
      options: {
        sigma:          { type: 'string' },
        method:         { type: 'string' },
        'channel-mode': { type: 'string' },
        'noise-floor':  { type: 'string' },
        bilateral:      { type: 'boolean' },
        workers:        { type: 'string' },
        watermark:      { type: 'boolean' },
        'watermark-seed': { type: 'string' },
        'watermark-strength': { type: 'string' },
        'watermark-payload': { type: 'string' },
        verbose:        { type: 'boolean', short: 'v' },
        help:           { type: 'boolean', short: 'h' },
      },
    });
    values = parsed.values as Record<string, string | boolean | undefined>;
    positionals = parsed.positionals;
  } catch (err) {
    console.error('affectively-deblur:', err instanceof Error ? err.message : String(err));
    console.error('');
    console.error(HELP_TEXT);
    return 1;
  }

  if (values.help) {
    process.stdout.write(HELP_TEXT);
    return 0;
  }

  const input = positionals[0];
  const output = positionals[1];

  if (!input || !output) {
    console.error('affectively-deblur: missing required <input> and <output> paths.');
    console.error('');
    console.error(HELP_TEXT);
    return 1;
  }

  let sigma: number | 'auto' | undefined;
  let method: 'wiener' | 'richardson-lucy' | undefined;
  let channelMode: 'rgb' | 'luminance' | undefined;
  let noiseFloor: number | 'auto' | undefined;
  let workers: number | undefined;
  let watermarkStrength: number | undefined;

  try {
    sigma = parseSigma(values.sigma as string | undefined);
    method = parseMethod(values.method as string | undefined);
    channelMode = parseChannelMode(values['channel-mode'] as string | undefined);
    noiseFloor = parseNoiseFloor(values['noise-floor'] as string | undefined);
    workers = parseInteger(values.workers as string | undefined, 'workers');
    watermarkStrength = parseWatermarkStrength(values['watermark-strength'] as string | undefined);
  } catch (err) {
    console.error('affectively-deblur:', err instanceof Error ? err.message : String(err));
    return 1;
  }

  const bilateral = values.bilateral === true;
  const watermark = values.watermark === true;
  const watermarkSeed = values['watermark-seed'] as string | undefined;
  const watermarkPayload = values['watermark-payload'] as string | undefined;
  const verbose = values.verbose === true;
  const invisibleWatermark: InvisibleWatermarkOptions | undefined = watermark
    ? {
        enabled: true,
        seed: watermarkSeed,
        strength: watermarkStrength,
        payload: watermarkPayload,
      }
    : undefined;

  const ext = extname(input).toLowerCase();

  try {
    if (IMAGE_EXTS.has(ext)) {
      if (verbose) {
        console.error(`[affectively-deblur] image: ${input} -> ${output}`);
        console.error(`[affectively-deblur] sigma=${sigma ?? 'default'} method=${method ?? 'default'} ` +
          `channel-mode=${channelMode ?? 'default'} noise-floor=${noiseFloor ?? 'default'} ` +
          `bilateral=${bilateral}`);
      }

      const imageOptions: DeblurAnyImageOptions = {};
      if (sigma !== undefined) imageOptions.sigma = sigma;
      if (method !== undefined) imageOptions.method = method;
      if (channelMode !== undefined) imageOptions.channelMode = channelMode;
      if (noiseFloor !== undefined) imageOptions.noiseFloor = noiseFloor;
      if (bilateral) imageOptions.bilateralPostpass = true;
      if (invisibleWatermark !== undefined) imageOptions.invisibleWatermark = invisibleWatermark;

      // deblur-any-image accepts a path string directly (string | Buffer | DecodedImageInput).
      const recovered = await deblurAnyImage(input, imageOptions);
      const png = await encodeRecoveredAsPNG(recovered);
      await writeFile(output, png);

      if (verbose) {
        console.error(`[affectively-deblur] wrote ${output} (${png.byteLength} bytes)`);
      }
      return 0;
    }

    if (VIDEO_EXTS.has(ext)) {
      if (verbose) {
        console.error(`[affectively-deblur] video: ${input} -> ${output}`);
        console.error(`[affectively-deblur] sigma=${sigma ?? 'default'} method=${method ?? 'default'} ` +
          `workers=${workers ?? 'default'} bilateral=${bilateral}`);
      }

      const videoDeblurOptions: DeblurAnyImageOptions = {};
      if (sigma !== undefined) videoDeblurOptions.sigma = sigma;
      if (method !== undefined) videoDeblurOptions.method = method;
      if (channelMode !== undefined) videoDeblurOptions.channelMode = channelMode;
      if (noiseFloor !== undefined) videoDeblurOptions.noiseFloor = noiseFloor;
      if (bilateral) videoDeblurOptions.bilateralPostpass = true;
      if (invisibleWatermark !== undefined) videoDeblurOptions.invisibleWatermark = invisibleWatermark;

      await deblurVideo({
        input,
        output,
        workers,
        options: videoDeblurOptions,
      });

      if (verbose) {
        console.error(`[affectively-deblur] video done: ${output}`);
      }
      return 0;
    }

    if (AUDIO_EXTS.has(ext)) {
      console.error('Audio CLI not yet wired — see audio-deblur.ts for programmatic API.');
      return 1;
    }

    console.error(`Unknown input extension: ${ext}`);
    console.error('Supported: .png .jpg .jpeg .webp .mp4 .mov .webm .avi .mkv .wav .flac');
    return 1;
  } catch (err) {
    console.error('affectively-deblur:', err instanceof Error ? err.message : String(err));
    if (verbose && err instanceof Error && err.stack) {
      console.error(err.stack);
    }
    return 1;
  }
}

// Auto-run guard: invoke main() when this file is executed directly,
// but stay inert when imported as a library module.
const isMainModule = (() => {
  try {
    const argv1 = process.argv[1];
    if (!argv1) return false;
    if (import.meta.url === `file://${argv1}`) return true;
    const tail = argv1.split('/').pop();
    if (tail && import.meta.url.endsWith('/' + tail)) return true;
    return false;
  } catch {
    return false;
  }
})();

if (isMainModule) {
  main(process.argv.slice(2))
    .then((code) => process.exit(code))
    .catch((err) => {
      console.error('affectively-deblur:', err);
      process.exit(1);
    });
}
