// @a0n/gnosis/distributed-inference-host/src/registry-shim.ts
import type { StationEndpoint } from './station';

export interface RegistryShimConfig {
  /** The URL of the Resonance Registry (e.g. `http://resonance-registry`) */
  registryUrl: string;
  /** The knot_id to ask the registry for. */
  knotId: string;
  /** The start layer for the requested station. */
  startLayer: number;
  /** The end layer for the requested station. */
  endLayer: number;
  /** Optional custom fetcher (useful for service bindings). Defaults to global fetch. */
  fetcher?: { fetch(req: Request | string, init?: RequestInit): Promise<Response> };
}

interface ResolveResponse {
  station: {
    station_endpoint: string;
  };
}

export class RegistryShim implements StationEndpoint {
  private readonly registryUrl: string;
  private readonly knotId: string;
  private readonly startLayer: number;
  private readonly endLayer: number;
  private readonly fetcher: { fetch(req: Request | string, init?: RequestInit): Promise<Response> };

  private cachedEndpoint: string | null = null;
  private resolvePromise: Promise<string> | null = null;

  constructor(config: RegistryShimConfig) {
    this.registryUrl = config.registryUrl.replace(/\/+$/, '');
    this.knotId = config.knotId;
    this.startLayer = config.startLayer;
    this.endLayer = config.endLayer;
    this.fetcher = config.fetcher ?? { fetch: (req, init) => fetch(req, init) };
  }

  /**
   * Lazily resolves the endpoint, then proxies the request to it.
   * If the proxy fetch fails (network error or 5xx), it invalidates the cache
   * and retries the resolution once.
   */
  async fetch(req: Request): Promise<Response> {
    const doFetch = async (): Promise<Response> => {
      const endpointUrl = await this.getEndpoint();
      
      // Clone the request but rewrite the URL to the target endpoint.
      // E.g. req.url might be "http://station-0/forward"
      // endpointUrl might be "https://worker-123.example.com"
      // We want to fetch "https://worker-123.example.com/forward"
      const url = new URL(req.url);
      const targetUrl = `${endpointUrl.replace(/\/+$/, '')}${url.pathname}${url.search}`;
      
      // Need to clone the request completely since we might consume the body.
      // But we can only read the body if we buffer it, which we must do if we want to retry.
      // However, Station sends ArrayBuffer payloads, so `req.clone()` will work properly.
      const proxyReq = new Request(targetUrl, {
        method: req.method,
        headers: req.headers,
        body: req.clone().body,
        duplex: 'half',
      } as RequestInit);

      return await this.fetcher.fetch(proxyReq);
    };

    try {
      const res = await doFetch();
      if (res.status >= 500 && res.status < 600) {
        /*
         * Audit-2026-04-25 (Bug 1): drain undrained 5xx body before retry.
         *
         * Why: Workers HTTP semantics treat a Response object as a live handle
         * to the underlying socket / readable stream. If we drop the reference
         * without consuming the body, the runtime keeps the connection in a
         * "held" state -- on Cloudflare Workers this can leak subrequests,
         * stall the IO gate, and on Node/undici it pins the agent's socket
         * pool. Calling `arrayBuffer()` (and discarding the result) is the
         * cheapest portable way to signal "we are done with this body" so the
         * runtime can release the socket before we open a fresh fetch on the
         * retry path. The try/catch swallows any drain error because the
         * response was already a failure -- we do not want a drain hiccup to
         * mask the real 5xx and prevent the retry.
         */
        try { await res.arrayBuffer(); } catch { /* drain best-effort */ }
        this.invalidate();
        const retry = await doFetch();
        if (retry.status >= 500 && retry.status < 600) {
          /*
           * Audit-2026-04-25 (Bug 2): drain terminal 5xx before returning.
           *
           * Why: when the *retry* is also 5xx we surface the failure to the
           * caller, but the caller is unlikely to consume the body (it will
           * inspect the status, log, and discard). An undrained body on the
           * returned Response leaves the underlying socket pinned for the
           * lifetime of GC, which on Workers is effectively forever inside
           * the request scope. Draining here -- and rebuilding a synthetic
           * Response that preserves status, statusText, and headers --
           * lets the runtime release the connection immediately while still
           * giving the caller everything they need to diagnose the failure.
           * We rebuild rather than mutate because Response bodies are
           * single-use: once `arrayBuffer()` has consumed the stream the
           * original Response is no longer readable, so we hand back a
           * fresh Response carrying the buffered bytes.
           */
          let body: ArrayBuffer | null = null;
          try { body = await retry.arrayBuffer(); } catch { /* drain best-effort */ }
          return new Response(body, {
            status: retry.status,
            statusText: retry.statusText,
            headers: retry.headers,
          });
        }
        return retry;
      }
      return res;
    } catch (err) {
      // Network failure -> invalidate and retry once.
      this.invalidate();
      return await doFetch();
    }
  }

  private async getEndpoint(): Promise<string> {
    if (this.cachedEndpoint) {
      return this.cachedEndpoint;
    }
    if (this.resolvePromise) {
      return this.resolvePromise;
    }

    /*
     * Audit-2026-04-25 (Bug 6): clear `resolvePromise` on rejection so the
     * next caller starts a fresh resolve.
     *
     * Why: we cache `this.resolvePromise` to coalesce concurrent in-flight
     * resolves -- multiple Station requests racing through `fetch()` should
     * share one registry round-trip, not stampede the registry with N
     * duplicate `/resolve` calls. The trap: a Promise's settled state is
     * sticky. If we cache a promise that later *rejects* (registry 5xx,
     * DNS failure, malformed response, etc.) and never clear the field,
     * every subsequent caller that finds `this.resolvePromise !== null`
     * will `await` the same already-rejected promise and re-throw the
     * stale error forever -- the shim becomes permanently wedged with no
     * recovery short of restarting the Worker.
     *
     * Fix: attach a `.catch` to the cached promise that nulls the field
     * on rejection (and rethrows so the awaiting caller still sees the
     * error). The `.catch` handler runs before any post-await continuation
     * because microtasks fire in attachment order, so a sequential caller
     * like `await shim.fetch(...)` followed by another `await shim.fetch(...)`
     * is guaranteed to see `this.resolvePromise === null` on the second
     * call and trigger a fresh resolve. We keep the success-path assignment
     * to `this.cachedEndpoint` inside the IIFE so concurrent awaiters
     * still share the same hit, and we explicitly null the in-flight slot
     * on success too (via the trailing assignment) so a future invalidate()
     * has a clean field to work with.
     *
     * We deliberately do *not* use a `finally` here: a `finally` clears
     * the slot on both success and failure, which is fine, but using a
     * separate `.catch` makes the rejection-handling intent explicit at
     * the read site -- future maintainers can see "yes, rejections clear
     * the cache" without having to reason about finally + await ordering.
     */
    const inflight = (async () => {
      const url = `${this.registryUrl}/resolve?knot_id=${encodeURIComponent(
        this.knotId
      )}&start=${this.startLayer}&end=${this.endLayer}`;

      const res = await this.fetcher.fetch(url, { method: 'GET' });
      if (!res.ok) {
        let detail = '';
        try {
          detail = await res.text();
        } catch {
          // ignore
        }
        throw new Error(
          `RegistryShim: Failed to resolve ${this.knotId} [${this.startLayer}, ${this.endLayer}): HTTP ${res.status} -- ${detail}`
        );
      }

      const data = (await res.json()) as ResolveResponse;
      if (!data?.station?.station_endpoint) {
        throw new Error('RegistryShim: Invalid response format from /resolve');
      }

      this.cachedEndpoint = data.station.station_endpoint;
      return this.cachedEndpoint;
    })();

    this.resolvePromise = inflight.then(
      (endpoint) => {
        // Success: clear the in-flight slot so future invalidate() / cache
        // misses operate on a clean field. `cachedEndpoint` is set inside
        // the IIFE, so subsequent calls short-circuit at the top of
        // getEndpoint() before consulting `resolvePromise`.
        this.resolvePromise = null;
        return endpoint;
      },
      (err) => {
        // Rejection: clear the in-flight slot BEFORE rethrowing so the
        // awaiting caller's `await` resumes into an error state, and any
        // subsequent caller starting a new `getEndpoint()` finds null and
        // attempts a fresh resolve. This is the core fix for Bug 6.
        this.resolvePromise = null;
        throw err;
      },
    );

    return this.resolvePromise;
  }

  private invalidate(): void {
    this.cachedEndpoint = null;
    this.resolvePromise = null;
  }
}
