// @a0n/gnosis/distributed-inference-host/src/sieve-algebra.ts
// TS runtime mirror of Gnosis/SubstrateSieve.lean. Implements the three
// operations of the sieve algebra (Substrate carrier + applyGrounding +
// executeStretch) over an arbitrary substrate type. The image-plane Sieve
// and audio Sieve are concrete instances; future ones (video, point cloud,
// volumetric) plug in by satisfying the SieveOps interface. The Lean
// theorems are stated with sorry — runtime callers can rely on the API
// without depending on the proofs landing. Tier bandwidths follow the
// observer-bandwidth tower: decagon(10) < pleroma(90) < tritonTriple(270).

export const TIER_BANDWIDTHS = {
  decagon: 10,
  pleroma: 90,
  tritonTriple: 270,
} as const;

export type Tier = keyof typeof TIER_BANDWIDTHS;
export const TIERS: readonly Tier[] = ['decagon', 'pleroma', 'tritonTriple'];

export interface Substrate<A> {
  carrier: A;
  /** Current observer bandwidth — one of the TIER_BANDWIDTHS values. */
  bandwidth: number;
}

export interface SieveOps<A> {
  /** Fold the substrate to the Pleromatic ground (Level 10). Idempotent. */
  applyGrounding: (sub: Substrate<A>) => Substrate<A>;
  /** Lift to a list of tier-stretched substrates, in tower-ascending order. */
  executeStretch: (sub: Substrate<A>) => Substrate<A>[];
}

/** Identity sieve — applyGrounding returns input; executeStretch returns
 *  three copies at increasing bandwidths. Useful as a baseline / sanity
 *  test that downstream consumers compose against. */
export function identitySieve<A>(): SieveOps<A> {
  return {
    applyGrounding: (sub) => ({ ...sub, bandwidth: TIER_BANDWIDTHS.decagon }),
    executeStretch: (sub) =>
      TIERS.map((t) => ({ ...sub, bandwidth: TIER_BANDWIDTHS[t] })),
  };
}

/** Verify the sieve laws empirically on a witness substrate. Returns the
 *  list of failing law names (empty if all pass). Used in tests and as a
 *  runtime sanity check when wiring new sieves. */
export function verifySieveLaws<A>(
  sieve: SieveOps<A>,
  witness: Substrate<A>,
  equals: (a: Substrate<A>, b: Substrate<A>) => boolean,
): readonly string[] {
  const failures: string[] = [];

  // Law 1: grounding_idempotent
  const g1 = sieve.applyGrounding(witness);
  const g2 = sieve.applyGrounding(g1);
  if (!equals(g1, g2)) failures.push('grounding_idempotent');

  // Law 2: stretch monotone bandwidth
  const stretch = sieve.executeStretch(witness);
  for (let i = 1; i < stretch.length; i++) {
    if (stretch[i]!.bandwidth < stretch[i - 1]!.bandwidth) {
      failures.push(`stretch_monotone_bandwidth@${i}`);
      break;
    }
  }

  // Law 3: closure_at_top — last tier is at TritonTriple bandwidth.
  if (stretch.length > 0) {
    const last = stretch[stretch.length - 1]!;
    if (last.bandwidth !== TIER_BANDWIDTHS.tritonTriple) {
      failures.push('closure_at_top');
    }
  }

  return failures;
}
