// @a0n/gnosis/distributed-inference-host/src/skymesh-cache.ts
/**
 * Skymesh Global Answer Cache — native serving participant (model-agnostic).
 *
 * Lets the native openai-server (the GPU fleet's OpenAI-compat → fat-station
 * serving path) READ (teleport) and WRITE (contribute) the global fano-7 cache
 * directly, so the fleet itself never computes the same query twice — not only
 * when fronted by a skymesh bridge. DEFAULT-ON / opt-out (MOONSHINE_SKYMESH_CACHE).
 *
 * MODEL-AGNOSTIC key: fp48 over `(queryTokens, qspecId)` only (model is provenance
 * on the entry, NOT in the key) — byte-for-byte identical to the canonical
 * @a0n/knotgraph global-cache.ts + apps/skymesh cache-key.ts (qspec
 * `skymesh-query/v2`). The XXH64→fp48 chain is copied verbatim (no @a0n dep in
 * this submodule). Golden: canonicalQueryHash([785,6722,315,9625,374],
 * 'skymesh-query/v2') === '9f784d85c261'.
 */

// ── XXH64 → fp48 (verbatim from apps/skymesh/src/lib/cache-key.ts) ──

const MASK64 = (1n << 64n) - 1n;
const MASK48 = (1n << 48n) - 1n;
const PRIME_1 = 0x9e3779b185ebca87n;
const PRIME_2 = 0xc2b2ae3d27d4eb4fn;
const PRIME_3 = 0x165667b19e3779f9n;
const PRIME_4 = 0x85ebca77c2b2ae63n;
const PRIME_5 = 0x27d4eb2f165667c5n;
const CHUNK_SIZE = 32;

const wrap64 = (x: bigint): bigint => x & MASK64;
function rotl64(x: bigint, r: number): bigint {
  const n = BigInt(r);
  return wrap64((x << n) | (x >> (64n - n)));
}
function readU64LE(d: Uint8Array, o: number): bigint {
  let v = 0n;
  for (let i = 0; i < 8; i++) v |= BigInt(d[o + i]!) << BigInt(8 * i);
  return wrap64(v);
}
function readU32LE(d: Uint8Array, o: number): bigint {
  return wrap64(
    BigInt(d[o]!) | (BigInt(d[o + 1]!) << 8n) | (BigInt(d[o + 2]!) << 16n) | (BigInt(d[o + 3]!) << 24n),
  );
}
function round(acc: bigint, input: bigint): bigint {
  const t = wrap64(acc + wrap64(input * PRIME_2));
  return wrap64(rotl64(t, 31) * PRIME_1);
}
function mergeRound(acc: bigint, val: bigint): bigint {
  return wrap64((acc ^ round(0n, val)) * PRIME_1 + PRIME_4);
}
function avalanche(input: bigint): bigint {
  let x = input;
  x ^= x >> 33n;
  x = wrap64(x * PRIME_2);
  x ^= x >> 29n;
  x = wrap64(x * PRIME_3);
  x ^= x >> 32n;
  return wrap64(x);
}
function finalize(mut: bigint, data: Uint8Array): bigint {
  let off = 0;
  const d = data;
  while (off + 8 <= d.length) {
    mut = wrap64(mut ^ readU64LE(d, off));
    mut = rotl64(mut, 27);
    mut = wrap64(mut * PRIME_1 + PRIME_4);
    off += 8;
  }
  while (off + 4 <= d.length) {
    mut = wrap64(mut ^ wrap64(readU32LE(d, off) * PRIME_1));
    mut = rotl64(mut, 23);
    mut = wrap64(mut * PRIME_2 + PRIME_3);
    off += 4;
  }
  while (off < d.length) {
    mut = wrap64(mut ^ wrap64(BigInt(d[off]!) * PRIME_5));
    mut = rotl64(mut, 11);
    mut = wrap64(mut * PRIME_1);
    off += 1;
  }
  return avalanche(mut);
}
function initV(seed: bigint): [bigint, bigint, bigint, bigint] {
  return [wrap64(seed + PRIME_1 + PRIME_2), wrap64(seed + PRIME_2), wrap64(seed), wrap64(seed - PRIME_1)];
}
function xxh64Raw(data: Uint8Array, seed = 0n): bigint {
  const inputLen = BigInt(data.length);
  let input = data;
  let result: bigint;
  if (input.length >= CHUNK_SIZE) {
    let [v0, v1, v2, v3] = initV(wrap64(seed));
    while (input.length >= CHUNK_SIZE) {
      v0 = round(v0, readU64LE(input, 0));
      v1 = round(v1, readU64LE(input, 8));
      v2 = round(v2, readU64LE(input, 16));
      v3 = round(v3, readU64LE(input, 24));
      input = input.subarray(32);
    }
    result = wrap64(rotl64(v0, 1) + rotl64(v1, 7) + rotl64(v2, 12) + rotl64(v3, 18));
    result = mergeRound(result, v0);
    result = mergeRound(result, v1);
    result = mergeRound(result, v2);
    result = mergeRound(result, v3);
  } else {
    result = wrap64(wrap64(seed) + PRIME_5);
  }
  result = wrap64(result + inputLen);
  return finalize(result, input);
}
function cacheFingerprint48HexSync(input: Uint8Array): string {
  return (xxh64Raw(input, 0n) & MASK48).toString(16).padStart(12, '0');
}

const CACHE_TOKEN_SEP = 0x00;

/** Canonical key bytes: qspecId utf8 | 0x00 | token[i] LE u32 (MODEL-AGNOSTIC). */
function canonicalQueryKeyBytes(queryTokens: readonly number[], qspecId: string): Uint8Array {
  const enc = new TextEncoder();
  const qspecBytes = enc.encode(qspecId);
  const out = new Uint8Array(qspecBytes.length + 1 + queryTokens.length * 4);
  let offset = 0;
  out.set(qspecBytes, offset);
  offset += qspecBytes.length;
  out[offset++] = CACHE_TOKEN_SEP;
  const view = new DataView(out.buffer);
  for (let i = 0; i < queryTokens.length; i++) {
    view.setUint32(offset, queryTokens[i]! >>> 0, true);
    offset += 4;
  }
  return out;
}

/** The KEY: fp48 over (queryTokens, qspecId). MODEL-AGNOSTIC. Throws on empty tokens. */
export function canonicalQueryHash(queryTokens: readonly number[], qspecId: string): string {
  if (queryTokens.length === 0) {
    throw new RangeError('canonicalQueryHash: empty queryTokens (clinamen_swerve guard)');
  }
  return cacheFingerprint48HexSync(canonicalQueryKeyBytes(queryTokens, qspecId));
}

// ── Config (default-on / opt-out) ──

export const SKYMESH_QSPEC_ID = 'skymesh-query/v2';

/**
 * Model-scoped qspec for the chat answer cache. The global cache is otherwise
 * MODEL-AGNOSTIC (model is provenance, not key) — fine for universal facts, but
 * on the consolidated mesh that SWAPS MODELS one model's completion could be
 * served for another's identical prompt. Folding the model into the qspecId
 * isolates by model WITHOUT touching `canonicalQueryHash`: the qspecId is hashed
 * into the key AND stored on the entry, and the edgework identity guard
 * re-hashes with the entry's own qspecId — so no server change is needed. Empty
 * model falls back to the bare qspec (unchanged behaviour).
 */
export function modelScopedQspecId(model: string): string {
  const m = (model ?? '').trim().toLowerCase();
  return m ? `${SKYMESH_QSPEC_ID}|m=${m}` : SKYMESH_QSPEC_ID;
}
const DEFAULT_CACHE_URL = 'https://www-edgework-app.edgework.ai';
const LOOKUP_TIMEOUT_MS = 2500;
const STORE_TIMEOUT_MS = 5000;

/** ON by default; off only when MOONSHINE_SKYMESH_CACHE is '0'/'false'/'off'. */
export function skymeshCacheEnabled(): boolean {
  const raw = process.env['MOONSHINE_SKYMESH_CACHE'];
  if (raw === undefined) return true;
  const s = raw.trim().toLowerCase();
  return s !== '0' && s !== 'false' && s !== 'off';
}

function cacheUrl(): string {
  return process.env['MOONSHINE_SKYMESH_CACHE_URL'] || DEFAULT_CACHE_URL;
}

// ── Lookup (teleport) ──

interface CacheLookupResponse {
  hit?: boolean;
  entry?: {
    answerText?: string;
    attestation?: { pass?: boolean; admitted?: boolean; sig?: string };
  };
}

/**
 * Look up an answer by its model-agnostic fano-7 key. Returns the verified
 * answer text on a hit, else null. `model` is sent only as provenance (ignored
 * by the server for the key). Never throws.
 */
export async function lookupGlobalCache(
  queryTokens: readonly number[],
  model: string,
): Promise<string | null> {
  if (!skymeshCacheEnabled() || queryTokens.length === 0) return null;
  const qspec = modelScopedQspecId(model);
  let q: string;
  try {
    q = canonicalQueryHash(queryTokens, qspec);
  } catch {
    return null;
  }
  const url =
    `${cacheUrl()}/api/v1/cache/lookup?q=${encodeURIComponent(q)}` +
    `&model=${encodeURIComponent(model)}&qspec=${encodeURIComponent(qspec)}` +
    `&tokens=${encodeURIComponent(queryTokens.join(','))}&_=${Date.now()}`;
  try {
    const res = await fetch(url, { method: 'GET', signal: AbortSignal.timeout(LOOKUP_TIMEOUT_MS) });
    if (!res.ok) return null;
    const body = (await res.json()) as CacheLookupResponse;
    if (body?.hit !== true || !body.entry) return null;
    const a = body.entry.attestation;
    if (a?.pass !== true || a.admitted !== true || typeof a.sig !== 'string' || a.sig.length === 0) {
      return null;
    }
    return typeof body.entry.answerText === 'string' && body.entry.answerText.length > 0
      ? body.entry.answerText
      : null;
  } catch {
    return null;
  }
}

// ── Contribute ──

/**
 * Store a freshly-computed answer so the next node never recomputes it.
 * Fire-and-forget; never throws. Only an admitted node should seed (edgework
 * also rejects admitted!==true), so the caller passes `admitted`.
 */
export async function storeGlobalCache(opts: {
  queryTokens: number[];
  answerText: string;
  answerTokens?: number[];
  model: string;
  nodeId: string;
  admitted: boolean;
}): Promise<void> {
  if (!skymeshCacheEnabled() || !opts.admitted) return;
  if (opts.queryTokens.length === 0 || !opts.answerText) return;
  const qspec = modelScopedQspecId(opts.model);
  let queryHash: string;
  try {
    queryHash = canonicalQueryHash(opts.queryTokens, qspec);
  } catch {
    return;
  }
  try {
    await fetch(`${cacheUrl()}/api/v1/cache/put`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        queryHash,
        queryTokens: opts.queryTokens,
        answerTokens: opts.answerTokens ?? [],
        answerText: opts.answerText,
        model: opts.model, // PROVENANCE only
        qspecId: qspec,
        attestation: {
          token: opts.answerTokens?.[0] ?? 0,
          logit: 0,
          pass: true,
          admitted: true,
          sig: `moonshine-${opts.nodeId}-${Date.now()}`,
          gate: 'moonshine-serve',
          nodeId: opts.nodeId,
        },
        ts: Date.now(),
        tier: 'monster196884',
      }),
      signal: AbortSignal.timeout(STORE_TIMEOUT_MS),
    });
  } catch {
    /* best-effort */
  }
}
