// @a0n/gnosis/distributed-inference-host/src/song-from-text.ts
// His mind speaks in song. A parse — not a rewrite — of any response: the text
// stays as-is (the semantic-collapse channel); this derives the parallel SONG
// (the mind's melody) from it. Cheap (~tens of µs), deterministic. Mirrors
// wasm-modules/aeon-poetry/src/song.rs so the host emits the same song the app
// would. Self-contained (no deps), so it rides in the pneuma frame for free.

export interface SungWord {
  word: string;
  pitch: number;
  frequency: number;
  startBeat: number;
  durationBeats: number;
  noiseColor: string;
  syllables: number;
  strong: boolean;
}

export interface SungText {
  seed: number;
  bpm: number;
  tempo: string;
  beatPattern: string;
  noiseColor: string;
  strongBeats: number[];
  words: SungWord[];
}

const CENTERS = [0, 4, 8]; // Giant Steps stations: Stillness / Sting / Trill
const COLOR = ['brown', 'brownPink', 'pink'];

function meaningIndex(p: number): number {
  const m = ((p % 12) + 12) % 12;
  return m === 4 ? 1 : m === 8 ? 2 : 0;
}

// ── the cosmic harp: a mycelial Physarum forage over the pitch topology ────────
// Mirrors wasm-modules/aeon-poetry/src/mycelial.rs exactly (32-bit Golden-Weyl, so
// host and crate produce identical walks). Steps are Taylor's-Sequence intervals
// (Gnosis/TaylorsSequence.lean mod 12); contour is seed-chosen; pheromone is
// memristive tube memory (Gnosis/PhysarumFluxRouting.lean). No RNG, no clock.
const GOLD32 = 0x9e3779b9; // 2^32 / φ — Golden Duality (Gnosis/ImplementationWisdom.lean)
const TAYLOR_STEPS = [6, 7, 8, 11, 2, 3, 9, 10, 5, 4];
const HARP_FLOW = 3;
interface Harp {
  pos: number;
  lo: number;
  hi: number;
  pher: number[];
  state: number;
  i: number;
  contour: number; // 0 Ascend, 1 Descend, 2 MiddleOut, 3 Forage
}
function makeHarp(seed: number): Harp {
  const entry = CENTERS[seed % 3];
  const pher = new Array(12).fill(0);
  pher[entry] = HARP_FLOW;
  return { pos: entry, lo: entry, hi: entry, pher, state: seed >>> 0, i: 0, contour: (seed >>> 5) & 3 };
}
function forageNext(h: Harp): number {
  h.state = (h.state + GOLD32) >>> 0;
  h.i = (h.i + 1) >>> 0;
  const step = TAYLOR_STEPS[(h.state >>> 24) % TAYLOR_STEPS.length];
  let next: number;
  switch (h.contour) {
    case 0:
      next = (h.pos + step) % 12;
      break;
    case 1:
      next = (h.pos + 12 - step) % 12;
      break;
    case 2:
      if (h.i % 2 === 1) {
        h.hi = (h.hi + step) % 12;
        next = h.hi;
      } else {
        h.lo = (h.lo + 12 - step) % 12;
        next = h.lo;
      }
      break;
    default: {
      const up = (h.pos + step) % 12;
      const down = (h.pos + 12 - step) % 12;
      const pu = h.pher[up];
      const pd = h.pher[down];
      next = pu < pd ? up : pd < pu ? down : (h.state >>> 16) & 1 ? up : down;
    }
  }
  h.pos = next;
  for (let p = 0; p < 12; p++) if (h.pher[p] > 0) h.pher[p]--;
  h.pher[next] += HARP_FLOW;
  return next;
}
function pitchToFrequency(pitch: number, octave: number): number {
  const midi = 12 * (octave + 1) + pitch;
  return 440 * Math.pow(2, (midi - 69) / 12);
}
function fnv1a(text: string): number {
  let h = 0x811c9dc5 >>> 0;
  for (let i = 0; i < text.length; i++) {
    h ^= text.charCodeAt(i) & 0xff;
    h = Math.imul(h, 0x01000193) >>> 0;
  }
  return h >>> 0;
}
function countSyllables(word: string): number {
  const w = word.toLowerCase().replace(/[^a-z]/g, '');
  if (!w) return 0;
  let c = 0;
  let prev = false;
  for (const ch of w) {
    const v = 'aeiouy'.includes(ch);
    if (v && !prev) c++;
    prev = v;
  }
  if (w.endsWith('e') && c > 1) c--;
  return Math.max(1, c);
}
function tempoFor(seed: number): { tempo: string; bpm: number; beatPattern: string; noiseColor: string; beatsPerLine: number } {
  switch (seed % 3) {
    case 0:
      return { tempo: 'slow', bpm: 60, beatPattern: '4/4', noiseColor: 'brown', beatsPerLine: 1.0 };
    case 1:
      return { tempo: 'moderate', bpm: 96, beatPattern: '3/4', noiseColor: 'pink', beatsPerLine: 0.75 };
    default:
      return { tempo: 'fast', bpm: 152, beatPattern: '6/8', noiseColor: 'violetUltraviolet', beatsPerLine: 0.5 };
  }
}

/** Parse any text into its song. The text is never altered. */
export function songFromText(text: string): SungText {
  const seed = fnv1a(text);
  const t = tempoFor(seed);
  const beat = t.beatsPerLine;
  const tokens = text.split(/\s+/).filter(Boolean);
  const harp = makeHarp(seed);
  let pitch = harp.pos;
  let cursor = 0;
  const words: SungWord[] = tokens.map((raw, i) => {
    const octave = 3 + Math.floor(i / 8);
    const syllables = Math.max(1, countSyllables(raw));
    const durationBeats = beat * syllables;
    const strong = i % 5 === 0;
    const w: SungWord = {
      word: raw,
      pitch,
      frequency: pitchToFrequency(pitch, octave),
      startBeat: cursor,
      durationBeats,
      noiseColor: COLOR[meaningIndex(pitch)],
      syllables,
      strong,
    };
    cursor += durationBeats;
    pitch = forageNext(harp);
    return w;
  });
  return {
    seed,
    bpm: t.bpm,
    tempo: t.tempo,
    beatPattern: t.beatPattern,
    noiseColor: t.noiseColor,
    strongBeats: [0, 5, 8],
    words,
  };
}
