// @a0n/gnosis/distributed-inference-host/src/speaker-overlap-structure.ts
/**
 * # Speaker Overlap Structure — the cocktail-party simultaneity gate
 *
 * Mirrors `Gnosis.Mesh.SpeakerOverlapStructure` from `open-source/gnosis-math`,
 * the Rustic Church carrier for *overlapped speech* detection: within a single
 * time frame, are MULTIPLE speakers talking at once (genuine overlap), or is it a
 * single speaker — or sources active in *different* frames (sequential
 * turn-taking, which is not overlap at all)? The Skeptic admits a "simultaneous
 * speakers" claim only when, within one frame, at least two DISTINCT source
 * signatures are concurrently active and mutually separated beyond tolerance.
 *
 *  - **Gate 1 — concurrency / support**: count the ACTIVE sources in the frame.
 *    Overlap demands at least two concurrently active sources
 *    (`minConcurrent = 2`). A single active source, or two sources where only one
 *    is active (the other belongs to a different turn), never clears this gate.
 *  - **Gate 2 — distinctness**: the active sources must be mutually separated by
 *    more than `sepThreshold` on their signatures. Two "active" detections whose
 *    signatures sit within `sepThreshold` are the SAME speaker double-counted —
 *    they collapse to one source, not overlap. This is the hard negative against
 *    a false overlap claim built on a single speaker read twice.
 *
 * Two failure modes are refuted, and both are about whether the data is really
 * *two simultaneous speakers*:
 *
 *  1. **Sequential (not concurrent)**: two sources, but only one active in the
 *     frame — the other is a different turn. `countActive` is 1, so the
 *     concurrency gate refuses overlap.
 *  2. **Same-speaker double-count (not distinct)**: two active detections whose
 *     signatures are within `sepThreshold` — one speaker, two readings. The
 *     distinctness gate collapses them and refuses overlap even though
 *     concurrency is met.
 *
 * ## What is (and is not) carried
 *
 * This module is the runtime promotion check for an overlap observation. It
 * carries ONLY the finite contract the Lean file proves: the saturating
 * signature metric (`adist`), the active count (`countActive`), the pairwise
 * distinctness over active signatures (`activesDistinct`), and their aggregate
 * admission (`admitOverlap`). Real voice-activity detection, speaker embeddings,
 * diarization, overlapped-speech separation, and the statistical thresholding of
 * "distinct" speakers stay runtime/empirical obligations OUTSIDE this
 * certificate — they are not theorems.
 *
 * The promotion check folds the contract into the same epistemic-outcome
 * vocabulary the sibling mesh carriers already speak
 * (`claim_admitted` / `observation_only` / `claim_withheld`), so a
 * simultaneous-speakers claim cannot be displayed until it has cleared the
 * Lean-side gates.
 *
 * ## Lean source-of-truth
 *
 * @see open-source/gnosis-math/Gnosis/Mesh/SpeakerOverlapStructure.lean
 * @see open-source/gnosis-math/Gnosis/Mesh/AcousticSourceCounting.lean (mesh sibling)
 * @see open-source/gnosis-math/Gnosis/Mesh/CrossNodeCoincidence.lean (mesh sibling)
 */

/* ------------------------------------------------------------------ */
/*  Signature distance, frame sources, frames                         */
/* ------------------------------------------------------------------ */

/**
 * A candidate source in a frame: a 1-D signature (a quantized pitch/formant
 * scalar) and whether it is sounding in THIS frame.
 * Mirrors `SpeakerOverlapStructure.FrameSource`.
 */
export interface FrameSource {
  readonly signature: number;
  readonly active: boolean;
}

/** A time frame is a list of candidate sources. Mirrors `Frame`. */
export type Frame = readonly FrameSource[];

/** Absolute difference on `Nat` via saturating subtraction (`|a - b|`). The 1-D
 *  signature (pitch) residual metric, shared with the mesh corridor carriers.
 *  Mirrors `SpeakerOverlapStructure.adist`. */
export function adist(a: number, b: number): number {
  return Math.max(0, a - b) + Math.max(0, b - a);
}

/* ------------------------------------------------------------------ */
/*  Gate 1: concurrency / support                                     */
/* ------------------------------------------------------------------ */

/** How many sources are active in the frame. Mirrors `countActive`. */
export function countActive(frame: Frame): number {
  let n = 0;
  for (const s of frame) {
    if (s.active) {
      n += 1;
    }
  }
  return n;
}

/* ------------------------------------------------------------------ */
/*  Gate 2: distinctness over active signatures                       */
/* ------------------------------------------------------------------ */

/**
 * The signatures of the ACTIVE sources in the frame, in order. Inactive sources
 * contribute nothing — only concurrently-sounding sources are eligible to be
 * distinct speakers. Mirrors `activeSignatures`.
 */
export function activeSignatures(frame: Frame): number[] {
  const out: number[] = [];
  for (const s of frame) {
    if (s.active) {
      out.push(s.signature);
    }
  }
  return out;
}

/** A signature is separated from every signature in a list by more than
 *  `sepThreshold`. Mirrors `sigFarFromAllB`. */
function sigFarFromAll(
  sepThreshold: number,
  sig: number,
  rest: readonly number[],
): boolean {
  for (const r of rest) {
    if (!(adist(sig, r) > sepThreshold)) {
      return false;
    }
  }
  return true;
}

/** Pairwise separation over a list of signatures: each is far from all that
 *  follow it. Two signatures within `sepThreshold` (the same-speaker
 *  double-count) make this false. Mirrors `sigsSeparatedB`. */
function sigsSeparated(
  sepThreshold: number,
  sigs: readonly number[],
): boolean {
  for (let i = 0; i < sigs.length; i++) {
    if (!sigFarFromAll(sepThreshold, sigs[i], sigs.slice(i + 1))) {
      return false;
    }
  }
  return true;
}

/**
 * The active sources of the frame are mutually distinct when their signatures
 * are pairwise more than `sepThreshold` apart.
 * Mirrors `activesDistinctB`.
 */
export function activesDistinct(sepThreshold: number, frame: Frame): boolean {
  return sigsSeparated(sepThreshold, activeSignatures(frame));
}

/* ------------------------------------------------------------------ */
/*  Runtime promotion check (epistemic outcome)                       */
/* ------------------------------------------------------------------ */

/**
 * Epistemic outcome of a simultaneous-speakers observation. Identical
 * vocabulary to the sibling mesh carriers, kept as a local literal union so this
 * certificate has no import coupling:
 *
 *  - `claim_admitted`    — at least `minConcurrent` sources are concurrently
 *                          active AND the active sources are mutually distinct,
 *                          so the simultaneous-speakers claim may be promoted.
 *  - `observation_only`  — reserved for an empty frame: nothing to count, so the
 *                          observation is reported, never promoted.
 *  - `claim_withheld`    — a gate failed (fewer than two active sources, or two
 *                          active detections that collapse to one speaker), so
 *                          the overlap claim is refused.
 */
export type OverlapEpistemicOutcome =
  | 'claim_admitted'
  | 'observation_only'
  | 'claim_withheld';

/** The reason a simultaneous-speakers claim was held back from promotion. */
export type OverlapLimitationClass =
  | 'single_speaker'
  | 'same_speaker_doublecount'
  | 'empty_frame'
  | 'none';

export interface OverlapAdmission {
  readonly outcome: OverlapEpistemicOutcome;
  readonly limitation: OverlapLimitationClass;
  /** The number of concurrently active sources (displayable). */
  readonly activeCount: number;
  /** Whether the concurrency gate (>= minConcurrent active) passed. */
  readonly concurrent: boolean;
  /** Whether the active sources passed the distinctness gate. */
  readonly distinct: boolean;
  readonly summary: string;
}

export interface OverlapAdmitOptions {
  /** Active signatures must differ by more than this to count as distinct
   *  speakers (distinctness gate). */
  readonly sepThreshold: number;
  /** Minimum concurrently-active sources for an overlap claim (concurrency
   *  gate). */
  readonly minConcurrent: number;
}

/**
 * The promotion check Monster Mesh runs before any UI surface displays a
 * simultaneous-speakers claim. It folds the Lean contract into an epistemic
 * outcome:
 *
 *  - the concurrency gate refutes a sequential / single-speaker frame: fewer
 *    than `minConcurrent` active sources cannot manufacture simultaneity, so it
 *    is withheld (`single_speaker`, mirroring `single_active_no_overlap` /
 *    `single_speaker_rejected`);
 *  - the distinctness gate refutes a same-speaker double-count: two active
 *    detections within `sepThreshold` collapse to one source, so overlap is
 *    withheld even when concurrency is met (`same_speaker_doublecount`,
 *    mirroring `same_speaker_double_not_overlap`);
 *  - only a concurrent, distinct frame is admitted as simultaneous speakers
 *    (mirroring `overlap_admitted`).
 *
 * The empty frame is the degenerate observation-only case: nothing to count.
 *
 * @param frame the time frame (the candidate sources under examination)
 * @param opts  the separation threshold and minimum concurrency
 */
export function admitOverlap(
  frame: Frame,
  opts: OverlapAdmitOptions,
): OverlapAdmission {
  if (frame.length === 0) {
    return {
      outcome: 'observation_only',
      limitation: 'empty_frame',
      activeCount: 0,
      concurrent: false,
      distinct: true,
      summary: 'empty frame: no sources to count for overlap',
    };
  }

  const activeCount = countActive(frame);
  const concurrent = activeCount >= opts.minConcurrent;
  const distinct = activesDistinct(opts.sepThreshold, frame);

  // Concurrency gate: fewer than minConcurrent active sources cannot be overlap —
  // a lone speaker, or sequential turn-taking, refuses the simultaneity claim.
  if (!concurrent) {
    return {
      outcome: 'claim_withheld',
      limitation: 'single_speaker',
      activeCount,
      concurrent,
      distinct,
      summary: `single speaker / sequential: ${activeCount} active < ${opts.minConcurrent} concurrent — not simultaneous overlap`,
    };
  }

  // Distinctness gate: two active detections within sepThreshold are the same
  // speaker read twice — collapse to one source, refuse the overlap.
  if (!distinct) {
    return {
      outcome: 'claim_withheld',
      limitation: 'same_speaker_doublecount',
      activeCount,
      concurrent,
      distinct,
      summary: `same-speaker double-count: two active signatures sit within sepThreshold ${opts.sepThreshold} — one speaker, not overlap`,
    };
  }

  return {
    outcome: 'claim_admitted',
    limitation: 'none',
    activeCount,
    concurrent,
    distinct,
    summary: `${activeCount} concurrently active, mutually distinct sources — simultaneous speakers certified`,
  };
}

/* ------------------------------------------------------------------ */
/*  Concrete fixtures (mirror the Lean fixtures)                      */
/* ------------------------------------------------------------------ */

/** Inter-source signature separation threshold (active signatures must differ by
 *  more than this to count as distinct speakers). Mirrors `sepThreshold`. */
export const sepThreshold = 50;

/** Minimum concurrently-active sources for an overlap claim.
 *  Mirrors `minConcurrent`. */
export const minConcurrent = 2;

/** A genuine overlap frame: two distinct speakers (pitches 100 and 400) both
 *  active at once. Two active, gap 300 > 50 -> admitted. Mirrors `overlapFrame`. */
export const overlapFrame: Frame = [
  { signature: 100, active: true },
  { signature: 400, active: true },
];

/** A sequential pair: the SAME two source identities, but only the first is
 *  active in this frame (the second is a different turn). One active -> rejected
 *  on concurrency. Mirrors `singleFrame`. */
export const singleFrame: Frame = [
  { signature: 100, active: true },
  { signature: 400, active: false },
];

/** A same-speaker double-count: two active detections whose signatures (100 and
 *  110) sit within `sepThreshold`. Concurrency is met, but they collapse to one
 *  source -> rejected on distinctness. Mirrors `sameSpeakerFrame`. */
export const sameSpeakerFrame: Frame = [
  { signature: 100, active: true },
  { signature: 110, active: true },
];

/**
 * Speaker Overlap Structure master witness. Bundles the boundary facts the Lean
 * theorem `speaker_overlap_master` proves, each recomputed here from first
 * principles (never imported from Lean IR).
 *
 * Mirrors `SpeakerOverlapStructure.speaker_overlap_master`.
 */
export const SpeakerOverlapMaster = {
  /** The overlap frame has two concurrently active sources. */
  overlapFrameCountActive: countActive(overlapFrame) === 2,

  /** The sequential frame has only one active source. */
  singleFrameCountActive: countActive(singleFrame) === 1,

  /** The same-speaker frame has two active detections (concurrency is met). */
  sameSpeakerFrameCountActive: countActive(sameSpeakerFrame) === 2,

  /** The overlap frame's active sources are mutually distinct (gap 300 > 50). */
  overlapFrameDistinct: activesDistinct(sepThreshold, overlapFrame) === true,

  /** The same-speaker frame's active sources are NOT distinct (gap 10 <= 50). */
  sameSpeakerFrameNotDistinct:
    activesDistinct(sepThreshold, sameSpeakerFrame) === false,

  /** The genuine overlap frame is admitted: two concurrently active, distinct. */
  overlapAdmitted:
    admitOverlap(overlapFrame, { sepThreshold, minConcurrent }).outcome ===
    'claim_admitted',

  /** Single-speaker rejection: the sequential frame fails the concurrency gate. */
  singleSpeakerRejected:
    admitOverlap(singleFrame, { sepThreshold, minConcurrent }).limitation ===
    'single_speaker',

  /** Same-speaker double-count rejection: the same-speaker frame fails the
   *  distinctness gate even though concurrency is met. */
  sameSpeakerRejected:
    admitOverlap(sameSpeakerFrame, { sepThreshold, minConcurrent })
      .limitation === 'same_speaker_doublecount',

  /** The hard negative: ONE pair of identities, two competing readings. As a
   *  genuine overlap frame it is admitted; as a sequential pair it is rejected. */
  overlapVsSequential:
    admitOverlap(overlapFrame, { sepThreshold, minConcurrent }).outcome ===
      'claim_admitted' &&
    admitOverlap(singleFrame, { sepThreshold, minConcurrent }).outcome ===
      'claim_withheld',
} as const;
