// @a0n/gnosis/distributed-inference-host/src/speculative-sampling.ts
/**
 * speculative-sampling.ts — training-free reference primitives for the two
 * speculative-decode axes described in docs/speculative-decoding-plan.md.
 *
 * Origin: assessment of NextLat (arXiv 2511.05963). NextLat's speedup is a
 * *trained* latent-rollout draft; we cannot train frozen GGUF weights, so this
 * module implements the two training-free substitutes:
 *
 *   Axis 1 — conformal layer early-exit (`pickExitLayer`). The codebase-native
 *     path: reuse the proven `2·τ ≤ γ` energy gate (rust `tukey_bowl.rs` /
 *     `conformal_gate.rs`) to fast-path out of transformer layers per token. A
 *     residual under half the decision margin provably cannot flip the argmax,
 *     so an admitted early exit needs NO draft/verify and NO KV rollback.
 *
 *   Axis 2 — lossless token-level accept/verify (`verifyDraftSpan`). The classic
 *     Leviathan/Chen speculative-sampling loop NextLat builds on: accept draft
 *     token i with prob min(1, p_i/q_i); on first reject, resample the residual
 *     max(0, p−q). Variable length: the caller proposes a variable-length span;
 *     this returns how much was accepted.
 *
 * Pure functions only — no transport, no model. The orchestrator injects
 * propose/verify callbacks (mirrors the stub-factory test style in this repo).
 */

// ─────────────────────────────────────────────────────────────────────────────
// Axis 1: conformal layer early-exit  (τ ≤ γ/2)
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Per-exit-layer calibration. `tauQ` is the conformal q-quantile `Q_q` of the
 * sup-norm residual between this exit layer's induced logits and the full-depth
 * logits, measured offline. Mirrors `ResidualCalibration::conformal_quantile`
 * in rust `conformal_gate.rs`. `layer` is the (0-based) layer index at which the
 * exit head would run.
 */
export interface ExitLayerCalibration {
  readonly layer: number;
  /** Conformal q-quantile of the early-exit→full-depth residual (sup-norm). */
  readonly tauQ: number;
  /** Coverage the quantile was taken at; certified flip-rate is ≤ 1 − q. */
  readonly q: number;
}

export interface ExitDecision {
  /** Layer to exit at, or null if no calibrated layer clears the gate. */
  readonly exitLayer: number | null;
  /** The decision margin γ that was tested. */
  readonly gamma: number;
  /** Certified upper bound on the flip rate at the chosen exit (1 − q), or 1. */
  readonly certifiedFlipRateLe: number;
}

/**
 * γ = top1 − top2 of the logits: the runner-up decision margin. A draft (or
 * early-exit) perturbation strictly below γ/2 in sup-norm cannot move the argmax
 * (the tukey_bowl energy bound). Returns 0 for degenerate (<2 entries) logits.
 */
export function topTwoMargin(logits: Float32Array | readonly number[]): number {
  let top1 = -Infinity;
  let top2 = -Infinity;
  for (let i = 0; i < logits.length; i++) {
    const v = logits[i]!;
    if (v > top1) {
      top2 = top1;
      top1 = v;
    } else if (v > top2) {
      top2 = v;
    }
  }
  if (!Number.isFinite(top2)) return 0;
  return top1 - top2;
}

/**
 * The `2·τ ≤ γ` admission predicate (rust `conformal_admit`). Admit iff the
 * calibrated residual quantile clears half the observed margin.
 */
export function admitsEarlyExit(tauQ: number, gamma: number): boolean {
  return gamma > 0 && Number.isFinite(tauQ) && 2 * tauQ <= gamma;
}

/**
 * Axis 1 decision. Given the current step's logit margin γ and the per-layer
 * calibrations (any order), pick the EARLIEST (shallowest) layer whose conformal
 * residual clears `2·Q_q ≤ γ`. This is the variable-length self-signal: large γ
 * (easy token) admits a very shallow exit; small γ (hard token) forces full depth.
 *
 * No verify pass is implied — the conformal certificate already bounds the
 * serve-time flip rate at `1 − q` for the chosen layer.
 */
export function pickExitLayer(
  gamma: number,
  calibrations: readonly ExitLayerCalibration[],
): ExitDecision {
  let best: ExitLayerCalibration | null = null;
  for (const c of calibrations) {
    if (!admitsEarlyExit(c.tauQ, gamma)) continue;
    if (best === null || c.layer < best.layer) best = c;
  }
  return {
    exitLayer: best?.layer ?? null,
    gamma,
    certifiedFlipRateLe: best ? 1 - best.q : 1,
  };
}

// ─────────────────────────────────────────────────────────────────────────────
// Axis 2: lossless token-level accept/verify  (Leviathan/Chen)
// ─────────────────────────────────────────────────────────────────────────────

/** One drafted step: the proposed token and the draft model's full distribution. */
export interface DraftStep {
  readonly token: number;
  /** Draft distribution q over the vocab at this step (proper probabilities). */
  readonly q: Float32Array | readonly number[];
}

export interface VerifyResult {
  /** Draft tokens accepted, in order. Length ∈ [0, draft.length]. */
  readonly acceptedTokens: number[];
  /**
   * The one extra token to emit after the accepted prefix: either the residual
   * resample (on the first rejection) or the bonus token sampled from the
   * target's distribution after a full-accept. Always present — the loop emits
   * at least one token per verify.
   */
  readonly correctedToken: number;
  /** Index of first rejected draft step, or draft.length if all accepted. */
  readonly rejectedAt: number;
}

/** Sample an index from an (unnormalized) nonneg weight vector given u ∈ [0,1). */
function sampleFrom(weights: readonly number[] | Float32Array, u: number): number {
  let total = 0;
  for (let i = 0; i < weights.length; i++) total += Math.max(0, weights[i]!);
  if (total <= 0) {
    // Degenerate residual (draft == target): fall back to argmax of the raw vec.
    let arg = 0;
    let max = -Infinity;
    for (let i = 0; i < weights.length; i++) {
      if (weights[i]! > max) {
        max = weights[i]!;
        arg = i;
      }
    }
    return arg;
  }
  const target = u * total;
  let acc = 0;
  for (let i = 0; i < weights.length; i++) {
    acc += Math.max(0, weights[i]!);
    if (acc > target) return i;
  }
  return weights.length - 1;
}

/**
 * Verify a variable-length draft span against the target distributions, losslessly.
 *
 * `targetDists[i]` is the target model's distribution p at draft step i (computed
 * in one batched verify forward over the speculative span). `targetDists` must
 * have length `draft.length + 1`: the trailing entry is the bonus distribution
 * used when every draft token is accepted.
 *
 * `rng()` must yield independent uniforms in [0,1). The output token distribution
 * is provably identical to sampling from the target model directly.
 */
export function verifyDraftSpan(
  draft: readonly DraftStep[],
  targetDists: readonly (Float32Array | readonly number[])[],
  rng: () => number,
): VerifyResult {
  if (targetDists.length !== draft.length + 1) {
    throw new Error(
      `verifyDraftSpan: targetDists must be draft.length+1 (${draft.length + 1}), ` +
        `got ${targetDists.length}`,
    );
  }
  const acceptedTokens: number[] = [];
  for (let i = 0; i < draft.length; i++) {
    const step = draft[i]!;
    const p = targetDists[i]!;
    const tok = step.token;
    const qTok = Math.max(0, step.q[tok] ?? 0);
    const pTok = Math.max(0, p[tok] ?? 0);
    const acceptProb = qTok <= 0 ? 1 : Math.min(1, pTok / qTok);
    if (rng() < acceptProb) {
      acceptedTokens.push(tok);
      continue;
    }
    // Rejected at i: resample the residual max(0, p − q), discard the rest.
    const residual = new Array(p.length);
    for (let v = 0; v < p.length; v++) {
      residual[v] = Math.max(0, (p[v] ?? 0) - (step.q[v] ?? 0));
    }
    return { acceptedTokens, correctedToken: sampleFrom(residual, rng()), rejectedAt: i };
  }
  // All accepted → bonus token from the trailing target distribution.
  const bonus = targetDists[draft.length]!;
  return {
    acceptedTokens,
    correctedToken: sampleFrom(bonus, rng()),
    rejectedAt: draft.length,
  };
}
