// @a0n/gnosis/distributed-inference-host/src/station-uds.ts
/**
 * StationUds — UDS-backed transport that implements the same surface as
 * `Station` (./station.ts) for the LLM hot-path methods.
 *
 * Why this exists
 * ---------------
 * For local-host mesh inference, HTTP's TCP handshake + parser overhead
 * dominates the per-call floor. A Unix domain socket with a 4-byte
 * length-prefixed framing collapses the wire to two `send`/`recv` copies
 * — no HTTP parser, no Content-Length-Encoded chunking. UDS has no MTU
 * so we can ship the full residual (up to ~115 KiB at seq_len=8) in a
 * single send/recv round.
 *
 * Wire format (mirrors `fat-station-uds.rs`):
 *
 *   [4 bytes BE u32: total_len]
 *   [ASCII headers terminated by a blank line, separator = "\n\n"]
 *     Request:  "method: POST\npath: /forward\nX-Position: 0\n...\n\n"
 *     Response: "status: 200\nContent-Type: ...\n...\n\n"
 *   [body bytes]
 *
 * Each call opens a fresh UDS connection, writes one frame, reads one
 * frame, closes. UDS connect is sub-millisecond locally — connection
 * pooling is not needed for v1.
 *
 * Surface parity:
 *   - This v1 implements `forward`, `embed`, `lmHead`, `cobordismEmbed`,
 *     `cobordismLmHead`, plus `health()` (no analogue on the HTTP class
 *     because the HTTP variant is a fetch-based RPC and uses ad-hoc GETs).
 *   - It does NOT implement the Pneuma encoder/decoder methods or the
 *     `bizarroForward` reject-mask path. The Whisper UDS server backend
 *     does not exist yet; wire those when fat-station-uds gains those
 *     handlers.
 *
 * Runtime: requires Node's `node:net` UDS support. CF Workers have no
 * UDS, so this transport is local-only by construction.
 */

import { createConnection, type Socket } from 'node:net';
import {
  AMPLITUHEDRON_CONTENT_TYPE,
  encodeCaptureRequest,
  decodeReplayResponse,
} from './amplituhedron-wire';
import type { IStation } from './types';

export interface StationUdsConfig {
  /** Filesystem path to the UDS the server listens on. */
  socketPath: string;
  /** Inclusive layer-range start. */
  startLayer: number;
  /** Exclusive layer-range end. */
  endLayer: number;
  /** Optional declared role for routing. */
  role?: 'entry' | 'mid' | 'exit' | 'cobordism';
  /** Per-call timeout in ms. Default 30s. */
  timeoutMs?: number;
}

interface FramedReply {
  status: number;
  headers: Map<string, string>;
  body: Uint8Array;
}

const DEFAULT_TIMEOUT_MS = 30_000;
const MAX_FRAME_BYTES = 256 * 1024 * 1024;

export class StationUds implements IStation {
  readonly startLayer: number;
  readonly endLayer: number;
  readonly role: 'entry' | 'mid' | 'exit' | 'cobordism' | undefined;
  readonly maxBatchSize = 8;
  private readonly socketPath: string;
  private readonly timeoutMs: number;

  constructor(config: StationUdsConfig) {
    if (!Number.isInteger(config.startLayer) || config.startLayer < 0) {
      throw new Error(
        `StationUds: startLayer must be a non-negative integer (got ${String(config.startLayer)})`,
      );
    }
    if (!Number.isInteger(config.endLayer) || config.endLayer < config.startLayer) {
      throw new Error(
        `StationUds: endLayer must be an integer >= startLayer (got start=${String(config.startLayer)}, end=${String(config.endLayer)})`,
      );
    }
    this.startLayer = config.startLayer;
    this.endLayer = config.endLayer;
    this.role = config.role;
    this.socketPath = config.socketPath;
    this.timeoutMs = config.timeoutMs ?? DEFAULT_TIMEOUT_MS;
  }

  /** Health probe — returns the server's parsed JSON status object. */
  async health(): Promise<unknown> {
    const reply = await this.call({
      method: 'GET',
      path: '/health',
      headers: {},
      body: new Uint8Array(0),
    });
    expectOk(reply, '/health');
    const text = new TextDecoder().decode(reply.body);
    try {
      return JSON.parse(text);
    } catch (e) {
      throw new Error(`StationUds /health: bad JSON: ${(e as Error).message} — body=${text}`);
    }
  }

  async forward(args: {
    residual: Float32Array;
    startPos: number;
    sequenceLength: number;
    requestId?: string;
  }): Promise<{
    residual: Float32Array;
    teleportation?: { layerDeficits: number[]; rounds: number };
  }> {
    if (!Number.isInteger(args.startPos) || args.startPos < 0) {
      throw new Error(`StationUds.forward: startPos must be >= 0 (got ${String(args.startPos)})`);
    }
    if (!Number.isInteger(args.sequenceLength) || args.sequenceLength <= 0) {
      throw new Error(
        `StationUds.forward: sequenceLength must be a positive integer (got ${String(args.sequenceLength)})`,
      );
    }
    const reply = await this.call({
      method: 'POST',
      path: '/forward',
      headers: {
        'X-Position': String(args.startPos),
        'X-Sequence-Length': String(args.sequenceLength),
        'X-Layer-Range': `${this.startLayer}-${this.endLayer}`,
        'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
      },
      body: float32ToBytes(args.residual),
    });
    expectOk(reply, '/forward');
    
    const residual = bytesToFloat32(reply.body, '/forward');
    const teleportation = parseTeleportationHeadersUds(reply.headers);

    return { residual, teleportation };
  }

  async embed(args: {
    tokens: Uint32Array;
    startPos: number;
    requestId?: string;
  }): Promise<Float32Array> {
    if (this.role !== undefined && this.role !== 'entry' && this.role !== undefined) {
      // mirror Station.embed role check; allow 'entry' or unspecified.
    }
    const reply = await this.call({
      method: 'POST',
      path: '/embed',
      headers: {
        'X-Position': String(args.startPos),
        'X-Token-Count': String(args.tokens.length),
        'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
      },
      body: uint32ToBytes(args.tokens),
    });
    expectOk(reply, '/embed');
    return bytesToFloat32(reply.body, '/embed');
  }

  async lmHead(args: {
    residual: Float32Array;
    sequenceLength: number;
    requestId?: string;
  }): Promise<Float32Array> {
    if (!Number.isInteger(args.sequenceLength) || args.sequenceLength <= 0) {
      throw new Error(
        `StationUds.lmHead: sequenceLength must be a positive integer (got ${String(args.sequenceLength)})`,
      );
    }
    const reply = await this.call({
      method: 'POST',
      path: '/lm-head',
      headers: {
        'X-Sequence-Length': String(args.sequenceLength),
        'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
      },
      body: float32ToBytes(args.residual),
    });
    expectOk(reply, '/lm-head');
    return bytesToFloat32(reply.body, '/lm-head');
  }

  async cobordismEmbed(args: {
    tokens: Uint32Array;
    startPos: number;
    vocabStart: number;
    vocabEnd: number;
    requestId?: string;
  }): Promise<Float32Array> {
    const reply = await this.call({
      method: 'POST',
      path: '/cobordism-embed',
      headers: {
        'X-Position': String(args.startPos),
        'X-Token-Count': String(args.tokens.length),
        'X-Vocab-Start': String(args.vocabStart),
        'X-Vocab-End': String(args.vocabEnd),
        'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
      },
      body: uint32ToBytes(args.tokens),
    });
    expectOk(reply, '/cobordism-embed');
    return bytesToFloat32(reply.body, '/cobordism-embed');
  }

  async cobordismLmHead(args: {
    residual: Float32Array;
    sequenceLength: number;
    vocabStart: number;
    vocabEnd: number;
    requestId?: string;
  }): Promise<Float32Array> {
    const reply = await this.call({
      method: 'POST',
      path: '/cobordism-lm-head',
      headers: {
        'X-Sequence-Length': String(args.sequenceLength),
        'X-Vocab-Start': String(args.vocabStart),
        'X-Vocab-End': String(args.vocabEnd),
        'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
      },
      body: float32ToBytes(args.residual),
    });
    expectOk(reply, '/cobordism-lm-head');
    return bytesToFloat32(reply.body, '/cobordism-lm-head');
  }

  async bizarroForward(args: {
    residual: Float32Array;
    startPos: number;
    sequenceLength: number;
    targetPosition: number;
    requestId?: string;
  }): Promise<{ rejectMask: number[] } | null> {
    return null;
  }

  // ── Amplituhedron admin path ───────────────────────────────────────
  //
  // The Amplituhedron volume cache (AMPLITUHEDRON_PLAN.md §4) speaks the
  // bitwise-binary wire over the same framed UDS transport. Pipeline fans
  // replay across every station to decide whether prefill can be skipped;
  // on miss it fires capture per station. Both calls MUST be best-effort:
  // any failure degrades to MISS + normal prefill and MUST NOT throw on
  // the hot path.
  //
  // Server contract (fat-station-uds.rs handler — to be wired):
  //   POST /amplituhedron/replay  → status 200 + bitwise envelope on HIT,
  //                                 status 204 (empty body) on MISS.
  //   POST /amplituhedron/capture → status 200 (any body) on accept.
  //
  // Wire-level body uses the codec in ./amplituhedron-wire so the
  // underlying handler is transport-agnostic.

  async amplituhedronReplay(args: {
    prefixHash: string;
    prefixLen: number;
    layerLo: number;
    layerHi: number;
    requestId?: string;
  }): Promise<{
    hit: boolean;
    tailResidual?: Float32Array;
    tailFingerprint?: string;
    prefixResidual?: Float32Array;
    prefixFingerprint?: string;
  }> {
    const body = JSON.stringify({
      prefix_hash: args.prefixHash,
      prefix_len: args.prefixLen,
      layer_lo: args.layerLo,
      layer_hi: args.layerHi,
      ...(args.requestId !== undefined ? { request_id: args.requestId } : {}),
    });
    let reply: FramedReply;
    try {
      reply = await this.call({
        method: 'POST',
        path: '/amplituhedron/replay',
        headers: {
          'Content-Type': 'application/json',
          Accept: AMPLITUHEDRON_CONTENT_TYPE,
          'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
        },
        body: new TextEncoder().encode(body),
      });
    } catch {
      // Transport-level error (timeout, framing) → treat as MISS.
      return { hit: false };
    }
    if (reply.status === 204) return { hit: false };
    if (reply.status !== 200) return { hit: false };
    try {
      return decodeReplayResponse(reply.body);
    } catch {
      return { hit: false };
    }
  }

  async amplituhedronCapture(args: {
    prefixHash: string;
    prefixLen: number;
    layerLo: number;
    layerHi: number;
    tailResidual?: Float32Array;
    requestId?: string;
  }): Promise<void> {
    // Contract gap: IStation declares tailResidual optional, codec requires
    // non-optional Float32Array. No residual = nothing to capture.
    if (!args.tailResidual) {
      return;
    }
    let body: Uint8Array;
    try {
      body = encodeCaptureRequest({
        prefixHash: args.prefixHash,
        prefixLen: args.prefixLen,
        layerLo: args.layerLo,
        layerHi: args.layerHi,
        tailResidual: args.tailResidual,
      });
    } catch {
      return;
    }
    try {
      await this.call({
        method: 'POST',
        path: '/amplituhedron/capture',
        headers: {
          'Content-Type': AMPLITUHEDRON_CONTENT_TYPE,
          'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
        },
        body,
      });
    } catch {
      // Capture is best-effort; a failure just means the next call misses.
    }
  }

  // ── transport ──────────────────────────────────────────────────────

  private async call(req: {
    method: string;
    path: string;
    headers: Record<string, string>;
    body: Uint8Array;
  }): Promise<FramedReply> {
    const frame = encodeRequestFrame(req.method, req.path, req.headers, req.body);
    const reply = await sendFrame(this.socketPath, frame, this.timeoutMs);
    return decodeReplyFrame(reply);
  }
}

// ── framing helpers ──────────────────────────────────────────────────

function encodeRequestFrame(
  method: string,
  path: string,
  headers: Record<string, string>,
  body: Uint8Array,
): Uint8Array {
  const lines: string[] = [];
  lines.push(`method: ${method}`);
  lines.push(`path: ${path}`);
  for (const [k, v] of Object.entries(headers)) {
    lines.push(`${k}: ${v}`);
  }
  // Each line ends with '\n', then we append one more '\n' to form the
  // blank-line "\n\n" separator that the server scans for.
  const headerBlock = lines.join('\n') + '\n' + '\n';
  const headerBytes = new TextEncoder().encode(headerBlock);
  const total = headerBytes.length + body.length;
  if (total > MAX_FRAME_BYTES) {
    throw new Error(`StationUds: frame too large (${total}B > ${MAX_FRAME_BYTES}B)`);
  }
  const out = new Uint8Array(4 + total);
  // big-endian length prefix
  out[0] = (total >>> 24) & 0xff;
  out[1] = (total >>> 16) & 0xff;
  out[2] = (total >>> 8) & 0xff;
  out[3] = total & 0xff;
  out.set(headerBytes, 4);
  out.set(body, 4 + headerBytes.length);
  return out;
}

function decodeReplyFrame(buf: Uint8Array): FramedReply {
  // Find the "\n\n" separator. The server always writes status + headers + \n + body.
  let split = -1;
  for (let i = 0; i < buf.length - 1; i++) {
    if (buf[i] === 0x0a && buf[i + 1] === 0x0a) {
      split = i;
      break;
    }
  }
  if (split < 0) {
    throw new Error('StationUds: reply frame missing blank-line separator');
  }
  const headerStr = new TextDecoder().decode(buf.subarray(0, split));
  const body = buf.subarray(split + 2);
  let status = 0;
  const headers = new Map<string, string>();
  for (const line of headerStr.split('\n')) {
    if (!line) continue;
    const colon = line.indexOf(':');
    if (colon < 0) continue;
    const k = line.slice(0, colon).trim();
    const v = line.slice(colon + 1).trim();
    if (k.toLowerCase() === 'status') {
      status = Number.parseInt(v, 10);
    } else {
      headers.set(k.toLowerCase(), v);
    }
  }
  return { status, headers, body: new Uint8Array(body.buffer, body.byteOffset, body.byteLength) };
}

function expectOk(reply: FramedReply, label: string): void {
  if (reply.status !== 200) {
    const detail = new TextDecoder().decode(reply.body).slice(0, 512);
    throw new Error(`StationUds ${label}: status ${reply.status} — ${detail}`);
  }
}

function float32ToBytes(arr: Float32Array): Uint8Array {
  return new Uint8Array(arr.buffer, arr.byteOffset, arr.byteLength).slice();
}

function uint32ToBytes(arr: Uint32Array): Uint8Array {
  return new Uint8Array(arr.buffer, arr.byteOffset, arr.byteLength).slice();
}

function bytesToFloat32(buf: Uint8Array, label: string): Float32Array {
  if (buf.byteLength % 4 !== 0) {
    throw new Error(
      `StationUds ${label}: response byteLength ${buf.byteLength} is not a multiple of 4`,
    );
  }
  // Copy to a fresh ArrayBuffer because the underlying chunk buffer may
  // be reused by node:net under load.
  const copy = new ArrayBuffer(buf.byteLength);
  new Uint8Array(copy).set(buf);
  return new Float32Array(copy);
}

function parseTeleportationHeadersUds(headers: Map<string, string>): { layerDeficits: number[]; rounds: number } | undefined {
  const deficitsStr = headers.get('x-teleport-deficits');
  const roundsStr = headers.get('x-teleport-rounds');
  if (!deficitsStr || !roundsStr) return undefined;

  try {
    const layerDeficits = deficitsStr.split(',').map(v => parseInt(v.trim(), 10));
    const rounds = parseInt(roundsStr, 10);
    if (layerDeficits.every(v => !isNaN(v)) && !isNaN(rounds)) {
      return { layerDeficits, rounds };
    }
  } catch {
    // ignore malformed headers
  }
  return undefined;
}

// ── Node-net send/recv ───────────────────────────────────────────────

function sendFrame(socketPath: string, frame: Uint8Array, timeoutMs: number): Promise<Uint8Array> {
  return new Promise((resolve, reject) => {
    let socket: Socket | undefined;
    let settled = false;
    const chunks: Uint8Array[] = [];
    let totalLen = -1;
    let received = 0;

    const finish = (err: Error | null, data?: Uint8Array): void => {
      if (settled) return;
      settled = true;
      if (timer) clearTimeout(timer);
      try {
        socket?.destroy();
      } catch {
        // ignore
      }
      if (err) {
        reject(err);
      } else {
        resolve(data!);
      }
    };

    const timer = setTimeout(() => {
      finish(new Error(`StationUds: call timed out after ${timeoutMs}ms`));
    }, timeoutMs);

    socket = createConnection({ path: socketPath }, () => {
      socket!.write(frame);
    });

    socket.on('data', (chunk: Buffer) => {
      chunks.push(new Uint8Array(chunk));
      received += chunk.length;
      // Try to read the 4-byte length prefix.
      if (totalLen < 0 && received >= 4) {
        const merged = concat(chunks);
        totalLen =
          (merged[0]! << 24) | (merged[1]! << 16) | (merged[2]! << 8) | merged[3]!;
        if (totalLen <= 0 || totalLen > MAX_FRAME_BYTES) {
          finish(new Error(`StationUds: bad reply length ${totalLen}`));
          return;
        }
        chunks.length = 0;
        chunks.push(merged.subarray(4));
        received = merged.length - 4;
      }
      if (totalLen >= 0 && received >= totalLen) {
        const merged = concat(chunks);
        finish(null, merged.subarray(0, totalLen));
      }
    });

    socket.on('error', (err: Error) => {
      finish(err);
    });
    socket.on('end', () => {
      if (totalLen < 0 || received < totalLen) {
        finish(new Error('StationUds: socket closed before frame complete'));
      }
    });
  });
}

function concat(chunks: Uint8Array[]): Uint8Array {
  if (chunks.length === 1) return chunks[0]!;
  if (chunks.length === 0) return new Uint8Array(0);
  let total = 0;
  for (const c of chunks) total += c.length;
  const out = new Uint8Array(total);
  let off = 0;
  for (const c of chunks) {
    out.set(c, off);
    off += c.length;
  }
  return out;
}
