// @a0n/gnosis/distributed-inference-host/src/station-flow.ts
/**
 * StationFlow — UDS-backed transport that speaks the canonical aeon-flow
 * 10-byte binary frame format.
 *
 * Why this exists
 * ---------------
 * `station-uds.ts` proved that UDS + a length-prefixed framing collapses
 * the per-call floor for local mesh inference. But its bespoke "ASCII
 * headers + blank line + body" format does not unify with the rest of
 * the aeon ecosystem. Every other aeon transport (the worker /flow
 * WebSocket, UDPFlowTransport, browser, etc.) already speaks the
 * `FlowCodec` 10-byte binary header.
 *
 * `StationFlow` mirrors `StationUds` byte-for-byte at the kernel surface
 * but exchanges aeon-flow frames. Pneuma's mesh-llm.ts can swap in
 * StationFlow without changing any kernel bookkeeping; the only
 * difference is the bytes on the wire.
 *
 * Wire format (mirror of fat-station-flow.rs)
 * -------------------------------------------
 * Each request and each response is exactly one aeon-flow frame:
 *
 *   Header (10 bytes, big-endian throughout):
 *     [0..1]  stream_id  u16   ROUTE OPCODE (see ROUTE_* constants below)
 *     [2..5]  sequence   u32   per-request id (echoed in response)
 *     [6]     flags      u8    bit 0x10 = FIN, bit 0x08 = VENT (error),
 *                              bit 0x20 = IS_RESPONSE, others reserved.
 *     [7..9]  length     u24   length of payload in bytes
 *
 *   Payload:
 *     [0..1]  meta_len   u16 BE  length of meta_json
 *     [2..]   meta_json  utf-8 JSON (start_pos, seq_len, layer_range, ...)
 *     [..]    body bytes (packed f32 / packed u32 / status JSON)
 *
 * Stream framing on UDS: the 10-byte aeon-flow header is wrapped in a
 * 4-byte big-endian length prefix so a stream-oriented socket can find
 * frame boundaries. On UDP this prefix is dropped — one datagram = one
 * aeon-flow frame.
 *
 * Surface parity:
 *   - Implements the full IStation surface (forward, embed, lmHead,
 *     cobordismEmbed, cobordismLmHead, plus health()) the same way
 *     StationUds does.
 *   - bizarroForward returns null (matches StationUds — that handler is
 *     not yet wired on the Rust side for this transport either).
 *
 * Runtime: requires Node `node:net`. CF Workers have no UDS, so this
 * transport is local-only by construction. Porting to UDP/WS reuses
 * `encodeFlowFrame` / `decodeFlowFrame` verbatim — only the socket and
 * (for UDS) the 4-byte length prefix change.
 */

import { createConnection, type Socket } from 'node:net';
import type { IStation } from './types';

// ── Aeon-flow constants (mirror open-source/aeon/src/flow/FlowCodec.ts) ──
export const FLOW_HEADER_SIZE = 10;
export const FLOW_MAX_PAYLOAD = 0xffffff;
export const FLAG_FIN = 0x10;
export const FLAG_VENT = 0x08;
export const FLAG_RESPONSE = 0x20;

// Route opcodes (mirror fat-station-flow.rs ROUTE_* constants).
export const ROUTE_HEALTH = 0x0001;
export const ROUTE_EMBED = 0x0002;
export const ROUTE_FORWARD = 0x0003;
export const ROUTE_LM_HEAD = 0x0004;
export const ROUTE_COBORDISM_EMBED = 0x0005;
export const ROUTE_COBORDISM_LM_HEAD = 0x0006;

const UDS_FRAMING_PREFIX_SIZE = 4;
const MAX_FRAME_BYTES = 256 * 1024 * 1024;
const DEFAULT_TIMEOUT_MS = 30_000;

export interface StationFlowConfig {
  /** Filesystem path to the UDS the server listens on. */
  socketPath: string;
  /** Inclusive layer-range start. */
  startLayer: number;
  /** Exclusive layer-range end. */
  endLayer: number;
  /** Optional declared role for routing. */
  role?: 'entry' | 'mid' | 'exit' | 'cobordism';
  /** Per-call timeout in ms. Default 30s. */
  timeoutMs?: number;
}

interface FlowFrameDecoded {
  streamId: number;
  sequence: number;
  flags: number;
  payload: Uint8Array;
}

interface ParsedReply {
  streamId: number;
  sequence: number;
  flags: number;
  meta: Record<string, unknown>;
  body: Uint8Array;
}

export class StationFlow implements IStation {
  readonly startLayer: number;
  readonly endLayer: number;
  readonly role: 'entry' | 'mid' | 'exit' | 'cobordism' | undefined;
  readonly maxBatchSize = 8;
  private readonly socketPath: string;
  private readonly timeoutMs: number;
  private nextSequence = 1;

  constructor(config: StationFlowConfig) {
    if (!Number.isInteger(config.startLayer) || config.startLayer < 0) {
      throw new Error(
        `StationFlow: startLayer must be a non-negative integer (got ${String(config.startLayer)})`,
      );
    }
    if (!Number.isInteger(config.endLayer) || config.endLayer < config.startLayer) {
      throw new Error(
        `StationFlow: endLayer must be an integer >= startLayer (got start=${String(config.startLayer)}, end=${String(config.endLayer)})`,
      );
    }
    this.startLayer = config.startLayer;
    this.endLayer = config.endLayer;
    this.role = config.role;
    this.socketPath = config.socketPath;
    this.timeoutMs = config.timeoutMs ?? DEFAULT_TIMEOUT_MS;
  }

  /** Health probe — returns the server's parsed JSON status object. */
  async health(): Promise<unknown> {
    const reply = await this.call(ROUTE_HEALTH, {}, new Uint8Array(0));
    expectOk(reply, 'health');
    // For health the response carries no body; the meta IS the status payload.
    return reply.meta;
  }

  async forward(args: {
    residual: Float32Array;
    startPos: number;
    sequenceLength: number;
    requestId?: string;
  }): Promise<{
    residual: Float32Array;
    teleportation?: { layerDeficits: number[]; rounds: number };
  }> {
    if (!Number.isInteger(args.startPos) || args.startPos < 0) {
      throw new Error(
        `StationFlow.forward: startPos must be >= 0 (got ${String(args.startPos)})`,
      );
    }
    if (!Number.isInteger(args.sequenceLength) || args.sequenceLength <= 0) {
      throw new Error(
        `StationFlow.forward: sequenceLength must be a positive integer (got ${String(args.sequenceLength)})`,
      );
    }
    const reply = await this.call(
      ROUTE_FORWARD,
      {
        start_pos: args.startPos,
        seq_len: args.sequenceLength,
        layer_range: `${this.startLayer}-${this.endLayer}`,
        request_id: args.requestId ?? `req-${Date.now()}`,
      },
      float32ToBytes(args.residual),
    );
    expectOk(reply, 'forward');
    const residual = bytesToFloat32(reply.body, 'forward');
    const teleportation = parseTeleportationMeta(reply.meta);
    return teleportation ? { residual, teleportation } : { residual };
  }

  async embed(args: {
    tokens: Uint32Array;
    startPos: number;
    requestId?: string;
  }): Promise<Float32Array> {
    const reply = await this.call(
      ROUTE_EMBED,
      {
        start_pos: args.startPos,
        token_count: args.tokens.length,
        request_id: args.requestId ?? `req-${Date.now()}`,
      },
      uint32ToBytes(args.tokens),
    );
    expectOk(reply, 'embed');
    return bytesToFloat32(reply.body, 'embed');
  }

  async lmHead(args: {
    residual: Float32Array;
    sequenceLength: number;
    requestId?: string;
  }): Promise<Float32Array> {
    if (!Number.isInteger(args.sequenceLength) || args.sequenceLength <= 0) {
      throw new Error(
        `StationFlow.lmHead: sequenceLength must be a positive integer (got ${String(args.sequenceLength)})`,
      );
    }
    const reply = await this.call(
      ROUTE_LM_HEAD,
      {
        seq_len: args.sequenceLength,
        request_id: args.requestId ?? `req-${Date.now()}`,
      },
      float32ToBytes(args.residual),
    );
    expectOk(reply, 'lm_head');
    return bytesToFloat32(reply.body, 'lm_head');
  }

  async cobordismEmbed(args: {
    tokens: Uint32Array;
    startPos: number;
    vocabStart: number;
    vocabEnd: number;
    requestId?: string;
  }): Promise<Float32Array> {
    const reply = await this.call(
      ROUTE_COBORDISM_EMBED,
      {
        start_pos: args.startPos,
        token_count: args.tokens.length,
        vocab_start: args.vocabStart,
        vocab_end: args.vocabEnd,
        request_id: args.requestId ?? `req-${Date.now()}`,
      },
      uint32ToBytes(args.tokens),
    );
    expectOk(reply, 'cobordism_embed');
    return bytesToFloat32(reply.body, 'cobordism_embed');
  }

  async cobordismLmHead(args: {
    residual: Float32Array;
    sequenceLength: number;
    vocabStart: number;
    vocabEnd: number;
    requestId?: string;
  }): Promise<Float32Array> {
    const reply = await this.call(
      ROUTE_COBORDISM_LM_HEAD,
      {
        seq_len: args.sequenceLength,
        vocab_start: args.vocabStart,
        vocab_end: args.vocabEnd,
        request_id: args.requestId ?? `req-${Date.now()}`,
      },
      float32ToBytes(args.residual),
    );
    expectOk(reply, 'cobordism_lm_head');
    return bytesToFloat32(reply.body, 'cobordism_lm_head');
  }

  async bizarroForward(_args: {
    residual: Float32Array;
    startPos: number;
    sequenceLength: number;
    targetPosition: number;
    requestId?: string;
  }): Promise<{ rejectMask: number[] } | null> {
    // Not yet wired on the Rust side for the flow transport (matches
    // StationUds). Returning null lets the caller fall back to the
    // non-bizarro path without erroring.
    return null;
  }

  // ── Amplituhedron admin path: intentionally NOT implemented ────────
  //
  // The Amplituhedron volume cache (AMPLITUHEDRON_PLAN.md §4) is a
  // JSON-shaped admin contract:
  //   POST /amplituhedron/replay  → 200 JSON envelope OR 204 No Content
  //   POST /amplituhedron/capture → 200
  //
  // StationFlow speaks the 10-byte aeon-flow binary frame format with
  // route-opcode dispatch (ROUTE_FORWARD, ROUTE_EMBED, etc.). There is
  // no ROUTE_AMPLITUHEDRON_* opcode allocated in fat-station-flow.rs, so
  // the server has no handler for these calls. Bolting a JSON path onto
  // an opcode-keyed binary transport would require a new wire convention
  // (e.g. ROUTE_AMPLITUHEDRON_REPLAY with the JSON ferried in `meta`)
  // and a matching Rust handler.
  //
  // Until both sides land, we omit `amplituhedronReplay` /
  // `amplituhedronCapture` here. The IStation interface marks both
  // `?optional`, and the coordinator (Pipeline.tryAmplituhedronReplay →
  // pipeline.ts:865) explicitly checks `if (!capable.amplituhedronReplay)`
  // before dispatching, so omission is the documented MISS path.

  // ── transport ──────────────────────────────────────────────────────

  private async call(
    streamId: number,
    meta: Record<string, unknown>,
    body: Uint8Array,
  ): Promise<ParsedReply> {
    const sequence = (this.nextSequence++ >>> 0);
    const payload = encodePayload(meta, body);
    const frame = encodeFlowFrame({
      streamId,
      sequence,
      flags: 0, // request: no FIN/VENT/RESPONSE bit; server sets RESPONSE|FIN.
      payload,
    });
    const wrapped = wrapForUds(frame);
    const reply = await sendFrame(this.socketPath, wrapped, this.timeoutMs);
    const decoded = decodeFlowFrame(reply);
    if ((decoded.flags & FLAG_RESPONSE) === 0) {
      throw new Error(
        `StationFlow: reply missing IS_RESPONSE flag (flags=0x${decoded.flags.toString(16)})`,
      );
    }
    if (decoded.streamId !== streamId) {
      throw new Error(
        `StationFlow: reply streamId 0x${decoded.streamId.toString(16)} != request 0x${streamId.toString(16)}`,
      );
    }
    if (decoded.sequence !== sequence) {
      throw new Error(
        `StationFlow: reply sequence ${decoded.sequence} != request ${sequence}`,
      );
    }
    const { meta: parsedMeta, body: parsedBody } = decodePayload(decoded.payload);
    return {
      streamId: decoded.streamId,
      sequence: decoded.sequence,
      flags: decoded.flags,
      meta: parsedMeta,
      body: parsedBody,
    };
  }
}

// ── aeon-flow codec ──────────────────────────────────────────────────

/**
 * Handles the gnosis encode Flow Frame workflow.
 */
export function encodeFlowFrame(frame: FlowFrameDecoded): Uint8Array {
  if (frame.payload.byteLength > FLOW_MAX_PAYLOAD) {
    throw new Error(
      `StationFlow: payload ${frame.payload.byteLength}B exceeds u24 max ${FLOW_MAX_PAYLOAD}B`,
    );
  }
  const out = new Uint8Array(FLOW_HEADER_SIZE + frame.payload.byteLength);
  out[0] = (frame.streamId >>> 8) & 0xff;
  out[1] = frame.streamId & 0xff;
  out[2] = (frame.sequence >>> 24) & 0xff;
  out[3] = (frame.sequence >>> 16) & 0xff;
  out[4] = (frame.sequence >>> 8) & 0xff;
  out[5] = frame.sequence & 0xff;
  out[6] = frame.flags & 0xff;
  const len = frame.payload.byteLength;
  out[7] = (len >> 16) & 0xff;
  out[8] = (len >> 8) & 0xff;
  out[9] = len & 0xff;
  out.set(frame.payload, FLOW_HEADER_SIZE);
  return out;
}

/**
 * Handles the gnosis decode Flow Frame workflow.
 */
export function decodeFlowFrame(bytes: Uint8Array): FlowFrameDecoded {
  if (bytes.byteLength < FLOW_HEADER_SIZE) {
    throw new Error(
      `StationFlow: frame ${bytes.byteLength}B < header ${FLOW_HEADER_SIZE}B`,
    );
  }
  const streamId = (bytes[0]! << 8) | bytes[1]!;
  const sequence =
    ((bytes[2]! * 0x1000000) + ((bytes[3]! << 16) | (bytes[4]! << 8) | bytes[5]!)) >>> 0;
  const flags = bytes[6]!;
  const length = (bytes[7]! << 16) | (bytes[8]! << 8) | bytes[9]!;
  if (bytes.byteLength < FLOW_HEADER_SIZE + length) {
    throw new Error(
      `StationFlow: truncated payload — declared=${length}B, available=${bytes.byteLength - FLOW_HEADER_SIZE}B`,
    );
  }
  return {
    streamId,
    sequence,
    flags,
    payload: bytes.subarray(FLOW_HEADER_SIZE, FLOW_HEADER_SIZE + length),
  };
}

function encodePayload(meta: Record<string, unknown>, body: Uint8Array): Uint8Array {
  const metaJson = JSON.stringify(meta);
  const metaBytes = new TextEncoder().encode(metaJson);
  if (metaBytes.byteLength > 0xffff) {
    throw new Error(
      `StationFlow: meta_json ${metaBytes.byteLength}B exceeds u16 max 65535`,
    );
  }
  const out = new Uint8Array(2 + metaBytes.byteLength + body.byteLength);
  out[0] = (metaBytes.byteLength >> 8) & 0xff;
  out[1] = metaBytes.byteLength & 0xff;
  out.set(metaBytes, 2);
  out.set(body, 2 + metaBytes.byteLength);
  return out;
}

function decodePayload(payload: Uint8Array): {
  meta: Record<string, unknown>;
  body: Uint8Array;
} {
  if (payload.byteLength < 2) {
    throw new Error('StationFlow: payload < 2B (no meta-length prefix)');
  }
  const metaLen = (payload[0]! << 8) | payload[1]!;
  if (payload.byteLength < 2 + metaLen) {
    throw new Error(
      `StationFlow: payload ${payload.byteLength}B < 2 + meta_len ${metaLen}`,
    );
  }
  const metaBytes = payload.subarray(2, 2 + metaLen);
  const metaJson = new TextDecoder().decode(metaBytes);
  let meta: Record<string, unknown>;
  try {
    meta = JSON.parse(metaJson) as Record<string, unknown>;
  } catch (e) {
    throw new Error(`StationFlow: meta json parse: ${(e as Error).message}`);
  }
  const body = payload.subarray(2 + metaLen);
  return { meta, body };
}

function wrapForUds(frame: Uint8Array): Uint8Array {
  // 4-byte big-endian length prefix so a stream socket can identify frame
  // boundaries. UDP transports drop this prefix entirely.
  if (frame.byteLength > MAX_FRAME_BYTES) {
    throw new Error(`StationFlow: frame too large (${frame.byteLength}B)`);
  }
  const wrapped = new Uint8Array(UDS_FRAMING_PREFIX_SIZE + frame.byteLength);
  wrapped[0] = (frame.byteLength >>> 24) & 0xff;
  wrapped[1] = (frame.byteLength >>> 16) & 0xff;
  wrapped[2] = (frame.byteLength >>> 8) & 0xff;
  wrapped[3] = frame.byteLength & 0xff;
  wrapped.set(frame, UDS_FRAMING_PREFIX_SIZE);
  return wrapped;
}

function expectOk(reply: ParsedReply, label: string): void {
  if ((reply.flags & FLAG_VENT) !== 0) {
    const errMsg =
      typeof reply.meta['error'] === 'string' ? (reply.meta['error'] as string) : 'unknown error';
    const status =
      typeof reply.meta['status'] === 'number' ? (reply.meta['status'] as number) : 500;
    throw new Error(`StationFlow ${label}: status ${status} — ${errMsg}`);
  }
  const status = reply.meta['status'];
  if (typeof status === 'number' && status !== 200) {
    throw new Error(`StationFlow ${label}: status ${status} — ${JSON.stringify(reply.meta)}`);
  }
}

function float32ToBytes(arr: Float32Array): Uint8Array {
  return new Uint8Array(arr.buffer, arr.byteOffset, arr.byteLength).slice();
}

function uint32ToBytes(arr: Uint32Array): Uint8Array {
  return new Uint8Array(arr.buffer, arr.byteOffset, arr.byteLength).slice();
}

function bytesToFloat32(buf: Uint8Array, label: string): Float32Array {
  if (buf.byteLength % 4 !== 0) {
    throw new Error(
      `StationFlow ${label}: response byteLength ${buf.byteLength} is not a multiple of 4`,
    );
  }
  // Copy to a fresh aligned ArrayBuffer because the underlying chunk
  // buffer may be reused by node:net under load AND because aeon-flow's
  // 10-byte header guarantees the f32 payload starts at offset 18 in the
  // raw frame — not 4-aligned. We copy here so callers always get an
  // independently-aligned Float32Array.
  const copy = new ArrayBuffer(buf.byteLength);
  new Uint8Array(copy).set(buf);
  return new Float32Array(copy);
}

function parseTeleportationMeta(
  meta: Record<string, unknown>,
): { layerDeficits: number[]; rounds: number } | undefined {
  const def = meta['teleport_deficits'];
  const rnd = meta['teleport_rounds'];
  if (!Array.isArray(def) || typeof rnd !== 'number') return undefined;
  const layerDeficits = def
    .map((v) => (typeof v === 'number' ? v : Number.parseInt(String(v), 10)))
    .filter((v) => Number.isFinite(v));
  if (layerDeficits.length !== def.length) return undefined;
  return { layerDeficits, rounds: rnd };
}

// ── Node-net send/recv ───────────────────────────────────────────────

function sendFrame(
  socketPath: string,
  frame: Uint8Array,
  timeoutMs: number,
): Promise<Uint8Array> {
  return new Promise((resolve, reject) => {
    let socket: Socket | undefined;
    let settled = false;
    const chunks: Uint8Array[] = [];
    let totalLen = -1;
    let received = 0;

    const finish = (err: Error | null, data?: Uint8Array): void => {
      if (settled) return;
      settled = true;
      if (timer) clearTimeout(timer);
      try {
        socket?.destroy();
      } catch {
        // ignore
      }
      if (err) reject(err);
      else resolve(data!);
    };

    const timer = setTimeout(() => {
      finish(new Error(`StationFlow: call timed out after ${timeoutMs}ms`));
    }, timeoutMs);

    socket = createConnection({ path: socketPath }, () => {
      socket!.write(frame);
    });

    socket.on('data', (chunk: Buffer) => {
      chunks.push(new Uint8Array(chunk));
      received += chunk.length;
      if (totalLen < 0 && received >= UDS_FRAMING_PREFIX_SIZE) {
        const merged = concat(chunks);
        totalLen =
          (merged[0]! << 24) | (merged[1]! << 16) | (merged[2]! << 8) | merged[3]!;
        if (totalLen <= 0 || totalLen > MAX_FRAME_BYTES) {
          finish(new Error(`StationFlow: bad reply length ${totalLen}`));
          return;
        }
        chunks.length = 0;
        chunks.push(merged.subarray(UDS_FRAMING_PREFIX_SIZE));
        received = merged.length - UDS_FRAMING_PREFIX_SIZE;
      }
      if (totalLen >= 0 && received >= totalLen) {
        const merged = concat(chunks);
        finish(null, merged.subarray(0, totalLen));
      }
    });

    socket.on('error', (err: Error) => finish(err));
    socket.on('end', () => {
      if (totalLen < 0 || received < totalLen) {
        finish(new Error('StationFlow: socket closed before frame complete'));
      }
    });
  });
}

function concat(chunks: Uint8Array[]): Uint8Array {
  if (chunks.length === 1) return chunks[0]!;
  if (chunks.length === 0) return new Uint8Array(0);
  let total = 0;
  for (const c of chunks) total += c.length;
  const out = new Uint8Array(total);
  let off = 0;
  for (const c of chunks) {
    out.set(c, off);
    off += c.length;
  }
  return out;
}
