// @a0n/gnosis/distributed-inference-host/src/station-uds-zerocopy.ts
/**
 * StationUdsZerocopy — SCM_RIGHTS / mmap-fd transport for the local
 * fat-station mesh. SKELETON v1: the wire spec, framing, and surface
 * are complete; the SCM_RIGHTS send/recv path requires native code
 * because Node's `node:net` module does NOT expose ancillary-data
 * (cmsg) on Unix sockets.
 *
 * Status legend
 * -------------
 *   IMPLEMENTED — pure-TS, runnable today.
 *   STUB        — surfaces the contract but throws at runtime; the
 *                 wiring is documented in-place so a follow-on agent
 *                 (or napi shim) can fill it in.
 *
 * Wire format (mirrors fat-station-uds-zerocopy.rs):
 *
 *   [4 bytes BE u32: total_len of header block]
 *   [ASCII headers terminated by "\n\n"]
 *      Request:  method, path, X-ZC-Body-Bytes, X-ZC-Body-Format, ...
 *      Response: status, X-ZC-Body-Bytes, X-ZC-Body-Format, ...
 *   <no inline body — payload travels as ancillary SCM_RIGHTS fd>
 *
 *   The ancillary fd points to an anonymous shared mmap region. Both
 *   sides mmap the same fd; bytes never leave kernel page cache.
 *
 * Why this is a SKELETON (and not a working client)
 * -------------------------------------------------
 * Node ships with `socket(2)`, `bind(2)`, `connect(2)`, `read(2)`,
 * `write(2)` exposed via `node:net` and `node:dgram`, but the
 * ancillary-data side of `sendmsg(2)` / `recvmsg(2)` — the part that
 * actually carries SCM_RIGHTS file descriptors — is not exposed.
 *
 * Three options, in order of effort, to make this real:
 *
 *   (a) `child_process.fork()` IPC channel: Node already uses
 *       SCM_RIGHTS internally to pass server sockets between parent
 *       and forked workers. The implementation lives in `src/handle_wrap.cc`
 *       but only the server-handle subset is exposed to JS. Could be
 *       wedged open with a tiny native addon that re-exports the
 *       cmsg path; the same machinery is already linked.
 *
 *   (b) NAPI shim in Rust (preferred): a 200-line `napi-rs` crate
 *       exposing two functions —
 *         sendmsgWithFd(unixSockFd: number, headerBytes: Buffer, payloadFd: number)
 *         recvmsgWithFd(unixSockFd: number, maxBytes: number) -> { headerBytes, payloadFd }
 *       — plus `memfdCreate(name, size)` / `shmOpenAnon(name, size)`
 *       and `mmapView(fd, length, prot)`. The fat-station-uds-zerocopy.rs
 *       binary already implements all the libc plumbing; the napi crate
 *       is essentially that file with `#[napi]` annotations.
 *
 *   (c) `node-ffi-napi` + a shim: the slowest path, dynamically loads
 *       libc symbols from JS and constructs `msghdr` / `cmsghdr` on
 *       the heap. Works without a build step but adds ~1 ms per call
 *       in FFI overhead — would defeat the entire purpose of the
 *       zero-copy transport.
 *
 * Recommendation: ship (b). The Rust↔Rust bench
 * (fat-station-uds-zerocopy --bench-zerocopy) proves the kernel-level
 * win is real (1.81× at seq_len=8, 3.27× at logits-sized payloads),
 * so the napi shim is well-motivated.
 *
 * Until then, this file:
 *   - documents the wire spec
 *   - implements the IStation surface so consuming code can be wired up
 *     (Pipeline.ts can `import { StationUdsZerocopy }` and reference it
 *     in the topology config without runtime errors at construction time)
 *   - throws a clear, actionable error when an actual call is made
 *
 * For comparison: the framed-copy variant lives in `./station-uds.ts`
 * and works today via plain `node:net`.
 */

import { createConnection, type Socket } from 'node:net';
import type { IStation } from './types';

export interface StationUdsZerocopyConfig {
  /** Filesystem path to the UDS the server listens on. */
  socketPath: string;
  /** Inclusive layer-range start. */
  startLayer: number;
  /** Exclusive layer-range end. */
  endLayer: number;
  /** Optional declared role for routing. */
  role?: 'entry' | 'mid' | 'exit' | 'cobordism';
  /** Per-call timeout in ms. Default 30s. */
  timeoutMs?: number;
  /**
   * Optional native shim that exposes SCM_RIGHTS sendmsg/recvmsg +
   * anonymous-shm fd creation. Pass `undefined` (default) to use the
   * skeleton (which throws). When the shim is provided, all surface
   * methods route through the zero-copy path.
   *
   * Expected shape — see top-of-file napi sketch (option b):
   *   {
   *     sendmsgWithFd(sockFd, headerBytes: Uint8Array, payloadFd: number | -1): void;
   *     recvmsgWithFd(sockFd, maxBytes): { headerBytes: Uint8Array; payloadFd: number };
   *     openAnonShm(name: string, byteLen: number): number;  // returns fd
   *     mmapRead(fd: number, byteLen: number): Uint8Array;   // shared view
   *     mmapWrite(fd: number, byteLen: number): Uint8Array;  // shared view
   *     close(fd: number): void;
   *   }
   */
  shim?: ZeroCopyShim;
}

/** See top-of-file for the napi shim sketch. */
export interface ZeroCopyShim {
  sendmsgWithFd(sockFd: number, headerBytes: Uint8Array, payloadFd: number): void;
  recvmsgWithFd(
    sockFd: number,
    maxBytes: number,
  ): { headerBytes: Uint8Array; payloadFd: number };
  openAnonShm(name: string, byteLen: number): number;
  mmapRead(fd: number, byteLen: number): Uint8Array;
  mmapWrite(fd: number, byteLen: number): Uint8Array;
  close(fd: number): void;
  /**
   * The Node `Socket.connect()` call returns a JS `Socket` whose
   * underlying file descriptor is intentionally hidden. The shim must
   * also provide a way to obtain that fd — typically by exposing a
   * `socketpair()` helper or by reading `socket._handle.fd` (private
   * API). In a production napi build this is a single libc::socket()
   * call; the JS-side `node:net` Socket is bypassed entirely.
   */
  connectUds(socketPath: string): number;
}

interface FramedReply {
  status: number;
  headers: Map<string, string>;
  bodyBytes: number;
  bodyFormat: 'f32' | 'u32' | 'u8' | 'empty';
  payloadFd: number; // -1 if no fd attached
}

const DEFAULT_TIMEOUT_MS = 30_000;
const MAX_HEADER_FRAME = 64 * 1024;

export class StationUdsZerocopy implements IStation {
  readonly startLayer: number;
  readonly endLayer: number;
  readonly role: 'entry' | 'mid' | 'exit' | 'cobordism' | undefined;
  private readonly socketPath: string;
  private readonly timeoutMs: number;
  private readonly shim: ZeroCopyShim | undefined;

  constructor(config: StationUdsZerocopyConfig) {
    if (!Number.isInteger(config.startLayer) || config.startLayer < 0) {
      throw new Error(
        `StationUdsZerocopy: startLayer must be a non-negative integer (got ${String(
          config.startLayer,
        )})`,
      );
    }
    if (!Number.isInteger(config.endLayer) || config.endLayer < config.startLayer) {
      throw new Error(
        `StationUdsZerocopy: endLayer must be an integer >= startLayer (got start=${String(
          config.startLayer,
        )}, end=${String(config.endLayer)})`,
      );
    }
    this.startLayer = config.startLayer;
    this.endLayer = config.endLayer;
    this.role = config.role;
    this.socketPath = config.socketPath;
    this.timeoutMs = config.timeoutMs ?? DEFAULT_TIMEOUT_MS;
    this.shim = config.shim;
  }

  // ── IStation surface ───────────────────────────────────────────────

  async forward(args: {
    residual: Float32Array;
    startPos: number;
    sequenceLength: number;
    requestId?: string;
  }): Promise<{ residual: Float32Array; teleportation?: { layerDeficits: number[]; rounds: number } }> {
    if (!Number.isInteger(args.startPos) || args.startPos < 0) {
      throw new Error(
        `StationUdsZerocopy.forward: startPos must be >= 0 (got ${String(args.startPos)})`,
      );
    }
    if (!Number.isInteger(args.sequenceLength) || args.sequenceLength <= 0) {
      throw new Error(
        `StationUdsZerocopy.forward: sequenceLength must be a positive integer (got ${String(
          args.sequenceLength,
        )})`,
      );
    }
    const reply = await this.callWithF32Body({
      method: 'POST',
      path: '/forward',
      headers: {
        'X-Position': String(args.startPos),
        'X-Sequence-Length': String(args.sequenceLength),
        'X-Layer-Range': `${this.startLayer}-${this.endLayer}`,
        'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
      },
      body: args.residual,
    });
    expectOk(reply, '/forward');
    const residual = this.fdToFloat32(reply, '/forward');
    const teleportation = parseTeleportationHeaders(reply.headers);
    return { residual, teleportation };
  }

  async embed(args: {
    tokens: Uint32Array;
    startPos: number;
    requestId?: string;
  }): Promise<Float32Array> {
    const reply = await this.callWithU32Body({
      method: 'POST',
      path: '/embed',
      headers: {
        'X-Position': String(args.startPos),
        'X-Token-Count': String(args.tokens.length),
        'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
      },
      body: args.tokens,
    });
    expectOk(reply, '/embed');
    return this.fdToFloat32(reply, '/embed');
  }

  async lmHead(args: {
    residual: Float32Array;
    sequenceLength: number;
    requestId?: string;
  }): Promise<Float32Array> {
    if (!Number.isInteger(args.sequenceLength) || args.sequenceLength <= 0) {
      throw new Error(
        `StationUdsZerocopy.lmHead: sequenceLength must be a positive integer (got ${String(
          args.sequenceLength,
        )})`,
      );
    }
    const reply = await this.callWithF32Body({
      method: 'POST',
      path: '/lm-head',
      headers: {
        'X-Sequence-Length': String(args.sequenceLength),
        'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
      },
      body: args.residual,
    });
    expectOk(reply, '/lm-head');
    return this.fdToFloat32(reply, '/lm-head');
  }

  async cobordismEmbed(args: {
    tokens: Uint32Array;
    startPos: number;
    vocabStart: number;
    vocabEnd: number;
    requestId?: string;
  }): Promise<Float32Array> {
    const reply = await this.callWithU32Body({
      method: 'POST',
      path: '/cobordism-embed',
      headers: {
        'X-Position': String(args.startPos),
        'X-Token-Count': String(args.tokens.length),
        'X-Vocab-Start': String(args.vocabStart),
        'X-Vocab-End': String(args.vocabEnd),
        'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
      },
      body: args.tokens,
    });
    expectOk(reply, '/cobordism-embed');
    return this.fdToFloat32(reply, '/cobordism-embed');
  }

  async cobordismLmHead(args: {
    residual: Float32Array;
    sequenceLength: number;
    vocabStart: number;
    vocabEnd: number;
    requestId?: string;
  }): Promise<Float32Array> {
    const reply = await this.callWithF32Body({
      method: 'POST',
      path: '/cobordism-lm-head',
      headers: {
        'X-Sequence-Length': String(args.sequenceLength),
        'X-Vocab-Start': String(args.vocabStart),
        'X-Vocab-End': String(args.vocabEnd),
        'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
      },
      body: args.residual,
    });
    expectOk(reply, '/cobordism-lm-head');
    return this.fdToFloat32(reply, '/cobordism-lm-head');
  }

  async bizarroForward(_args: {
    residual: Float32Array;
    startPos: number;
    sequenceLength: number;
    targetPosition: number;
    requestId?: string;
  }): Promise<{ rejectMask: number[] } | null> {
    return null;
  }

  // ── Amplituhedron admin path: intentionally NOT implemented ────────
  //
  // The zero-copy transport's entire reason for existing is to fold the
  // body into a `mmap`'d shm region passed by SCM_RIGHTS — the wire is
  // optimized for large f32 residuals. The Amplituhedron admin contract
  // ships JSON envelopes (potentially with `tail_residual` inline as a
  // number[]) and either 200+JSON or 204+empty replies. Marshalling a
  // JSON blob through the fd-passing path adds latency in the WRONG
  // direction and gains us nothing over the framed-copy UDS path that
  // StationUds (./station-uds.ts) already implements.
  //
  // Operators wanting Amplituhedron replay on a zero-copy mesh should
  // run StationUds alongside StationUdsZerocopy on the same UDS — the
  // server-side handler can dispatch on path regardless of whether the
  // payload arrived inline or as an fd. IStation marks both methods
  // `?optional` and Pipeline guards with `if (!capable.amplituhedronReplay)`,
  // so omission here degrades cleanly to MISS + normal prefill.

  // ── transport core ─────────────────────────────────────────────────

  private async callWithF32Body(req: {
    method: string;
    path: string;
    headers: Record<string, string>;
    body: Float32Array;
  }): Promise<FramedReply> {
    const bytes = new Uint8Array(req.body.buffer, req.body.byteOffset, req.body.byteLength);
    return this.callWithBody(req.method, req.path, req.headers, bytes, 'f32');
  }

  private async callWithU32Body(req: {
    method: string;
    path: string;
    headers: Record<string, string>;
    body: Uint32Array;
  }): Promise<FramedReply> {
    const bytes = new Uint8Array(req.body.buffer, req.body.byteOffset, req.body.byteLength);
    return this.callWithBody(req.method, req.path, req.headers, bytes, 'u32');
  }

  /**
   * The full SCM_RIGHTS round-trip. Skeleton — relies on `this.shim`.
   * When shim is undefined, throws a descriptive error so callers know
   * exactly what's missing.
   */
  private async callWithBody(
    method: string,
    path: string,
    headers: Record<string, string>,
    body: Uint8Array,
    bodyFormat: 'f32' | 'u32',
  ): Promise<FramedReply> {
    if (this.shim === undefined) {
      throw new Error(
        `StationUdsZerocopy: native SCM_RIGHTS shim is not configured. ` +
          `node:net does not expose sendmsg ancillary data, so this transport ` +
          `cannot run from pure JS. Either:\n` +
          `  (a) build a napi shim per the sketch in station-uds-zerocopy.ts,\n` +
          `  (b) fall back to StationUds (./station-uds.ts) which uses the\n` +
          `      framed-copy UDS path — slightly slower but pure-Node.\n` +
          `Call: ${method} ${path} (body=${body.byteLength}B ${bodyFormat})`,
      );
    }

    // ── 1. allocate anon shm fd, copy body in (this is the ONE memcpy
    //       we do on the send path; the kernel→peer copy is what we
    //       eliminate).
    const sockFd = this.shim.connectUds(this.socketPath);
    const payloadFd = this.shim.openAnonShm('zc-req', Math.max(body.byteLength, 1));
    try {
      if (body.byteLength > 0) {
        const dst = this.shim.mmapWrite(payloadFd, body.byteLength);
        dst.set(body);
      }

      // ── 2. build the ASCII control frame (no body bytes).
      const headerBytes = encodeRequestHeaderFrame(method, path, headers, body.byteLength, bodyFormat);

      // ── 3. sendmsg with the payload fd as ancillary data.
      this.shim.sendmsgWithFd(sockFd, headerBytes, payloadFd);

      // ── 4. recvmsg the response. Server sends a small data frame +
      //       a fresh shm fd carrying the response.
      const { headerBytes: respHeaderBytes, payloadFd: respFd } = this.shim.recvmsgWithFd(
        sockFd,
        MAX_HEADER_FRAME,
      );
      const reply = decodeReplyFrame(respHeaderBytes, respFd);
      return reply;
    } finally {
      this.shim.close(payloadFd);
      this.shim.close(sockFd);
    }
  }

  /**
   * Read the response payload as Float32Array. The shim's `mmapRead`
   * returns a shared view onto kernel page cache; we slice into a
   * fresh ArrayBuffer because the caller will outlive the mmap (we
   * close the fd in the finally block).
   *
   * A future optimization: thread the mmap through to the consumer
   * and let it borrow the view until done. That would require Pipeline.ts
   * to release residuals deterministically — additive contract change.
   */
  private fdToFloat32(reply: FramedReply, label: string): Float32Array {
    if (this.shim === undefined) {
      throw new Error(`StationUdsZerocopy.${label}: shim is undefined; cannot read response fd`);
    }
    if (reply.bodyBytes === 0) {
      return new Float32Array(0);
    }
    if (reply.bodyBytes % 4 !== 0) {
      throw new Error(
        `StationUdsZerocopy ${label}: response bodyBytes ${reply.bodyBytes} not a multiple of 4`,
      );
    }
    const view = this.shim.mmapRead(reply.payloadFd, reply.bodyBytes);
    // Copy into an owned ArrayBuffer — see comment above.
    const out = new ArrayBuffer(view.byteLength);
    new Uint8Array(out).set(view);
    this.shim.close(reply.payloadFd);
    return new Float32Array(out);
  }
}

// ── framing helpers ──────────────────────────────────────────────────

function encodeRequestHeaderFrame(
  method: string,
  path: string,
  headers: Record<string, string>,
  bodyBytes: number,
  bodyFormat: 'f32' | 'u32' | 'empty',
): Uint8Array {
  const lines: string[] = [];
  lines.push(`method: ${method}`);
  lines.push(`path: ${path}`);
  lines.push(`X-ZC-Body-Bytes: ${bodyBytes}`);
  lines.push(`X-ZC-Body-Format: ${bodyFormat}`);
  for (const [k, v] of Object.entries(headers)) {
    lines.push(`${k}: ${v}`);
  }
  const headerBlock = lines.join('\n') + '\n' + '\n';
  const headerBytes = new TextEncoder().encode(headerBlock);
  if (headerBytes.length > MAX_HEADER_FRAME) {
    throw new Error(
      `StationUdsZerocopy: header frame too large (${headerBytes.length}B > ${MAX_HEADER_FRAME}B)`,
    );
  }
  const out = new Uint8Array(4 + headerBytes.length);
  const total = headerBytes.length;
  out[0] = (total >>> 24) & 0xff;
  out[1] = (total >>> 16) & 0xff;
  out[2] = (total >>> 8) & 0xff;
  out[3] = total & 0xff;
  out.set(headerBytes, 4);
  return out;
}

function decodeReplyFrame(buf: Uint8Array, payloadFd: number): FramedReply {
  if (buf.length < 4) {
    throw new Error('StationUdsZerocopy: reply shorter than 4-byte length prefix');
  }
  const total = ((buf[0]! << 24) | (buf[1]! << 16) | (buf[2]! << 8) | buf[3]!) >>> 0;
  if (total === 0 || total > MAX_HEADER_FRAME) {
    throw new Error(`StationUdsZerocopy: reply total_len ${total} out of bounds`);
  }
  const frame = buf.subarray(4, 4 + total);
  let split = -1;
  for (let i = 0; i < frame.length - 1; i++) {
    if (frame[i] === 0x0a && frame[i + 1] === 0x0a) {
      split = i;
      break;
    }
  }
  if (split < 0) {
    throw new Error('StationUdsZerocopy: reply missing blank-line separator');
  }
  const headerStr = new TextDecoder().decode(frame.subarray(0, split));
  let status = 0;
  let bodyBytes = 0;
  let bodyFormat: 'f32' | 'u32' | 'u8' | 'empty' = 'empty';
  const headers = new Map<string, string>();
  for (const line of headerStr.split('\n')) {
    if (!line) continue;
    const colon = line.indexOf(':');
    if (colon < 0) continue;
    const k = line.slice(0, colon).trim();
    const v = line.slice(colon + 1).trim();
    const lk = k.toLowerCase();
    if (lk === 'status') {
      status = Number.parseInt(v, 10);
    } else if (lk === 'x-zc-body-bytes') {
      bodyBytes = Number.parseInt(v, 10);
    } else if (lk === 'x-zc-body-format') {
      const lv = v.toLowerCase();
      if (lv === 'f32' || lv === 'u32' || lv === 'u8' || lv === 'empty') {
        bodyFormat = lv as typeof bodyFormat;
      }
    } else {
      headers.set(lk, v);
    }
  }
  return { status, headers, bodyBytes, bodyFormat, payloadFd };
}

function expectOk(reply: FramedReply, label: string): void {
  if (reply.status !== 200) {
    throw new Error(`StationUdsZerocopy ${label}: status ${reply.status}`);
  }
}

function parseTeleportationHeaders(
  headers: Map<string, string>,
): { layerDeficits: number[]; rounds: number } | undefined {
  const deficitsStr = headers.get('x-teleport-deficits');
  const roundsStr = headers.get('x-teleport-rounds');
  if (!deficitsStr || !roundsStr) return undefined;
  try {
    const layerDeficits = deficitsStr.split(',').map((v) => parseInt(v.trim(), 10));
    const rounds = parseInt(roundsStr, 10);
    if (layerDeficits.every((v) => !isNaN(v)) && !isNaN(rounds)) {
      return { layerDeficits, rounds };
    }
  } catch {
    // ignore
  }
  return undefined;
}

// ── pure-Node fallback escape hatch ──────────────────────────────────

/**
 * When SCM_RIGHTS is unavailable, you can still test reachability of
 * the zero-copy server by sending the framed `/health` request — it
 * returns a small JSON body in an attached fd. WITHOUT a shim there's
 * no way to read that fd from JS, so `health()` throws a structured
 * error instead. This stub exists so callers can `instanceof`-check
 * in topology setup without a runtime crash.
 */
export async function probeReachable(socketPath: string): Promise<void> {
  return new Promise((resolve, reject) => {
    let socket: Socket | undefined;
    const timer = setTimeout(() => {
      try {
        socket?.destroy();
      } catch {
        // ignore
      }
      reject(new Error(`StationUdsZerocopy: connect to ${socketPath} timed out`));
    }, 1000);
    socket = createConnection({ path: socketPath }, () => {
      clearTimeout(timer);
      try {
        socket?.destroy();
      } catch {
        // ignore
      }
      resolve();
    });
    socket.on('error', (err: Error) => {
      clearTimeout(timer);
      reject(err);
    });
  });
}
