// @a0n/gnosis/distributed-inference-host/src/style-transfer-monocular-host.ts
/**
 * Host-side monocular-style admission (cheap proxies; not a full port of
 * `build_monocular_cue_maps`). Produces a per-pixel weight in `[0,1]` where
 * lower means "reduce style admission" (e.g. strong chroma variance as haze /
 * instability proxy).
 */

export function hostMonocularAdmissionFromRgb(
  rgb: Float32Array,
  width: number,
  height: number,
  channels: 1 | 3,
): Float32Array {
  const n = width * height;
  const out = new Float32Array(n);
  const luma = channels === 1 ? null : new Float32Array(n);
  if (channels === 3) {
    for (let i = 0; i < n; i++) {
      const r = rgb[i * 3]!;
      const g = rgb[i * 3 + 1]!;
      const b = rgb[i * 3 + 2]!;
      luma![i] = 0.2126 * r + 0.7152 * g + 0.0722 * b;
    }
  }
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      const p = y * width + x;
      if (channels === 1) {
        out[p] = 1;
        continue;
      }
      const r = rgb[p * 3]!;
      const g = rgb[p * 3 + 1]!;
      const b = rgb[p * 3 + 2]!;
      const chroma = (Math.abs(r - g) + Math.abs(g - b) + Math.abs(b - r)) / 3;
      const aerial = clamp01(1 - chroma * 1.8);
      let tex = 1;
      if (x > 0 && x + 1 < width && y > 0 && y + 1 < height) {
        const c = luma![p]!;
        const nbrs = [
          luma![(y - 1) * width + x]!,
          luma![(y + 1) * width + x]!,
          luma![y * width + (x - 1)]!,
          luma![y * width + (x + 1)]!,
        ];
        let v = 0;
        for (const z of nbrs) v += (z - c) * (z - c);
        tex = clamp01(1 - Math.sqrt(v / 4) * 3);
      }
      out[p] = clamp01(0.45 * aerial + 0.55 * tex);
    }
  }
  return out;
}

function clamp01(v: number): number {
  if (v <= 0) return 0;
  if (v >= 1) return 1;
  return v;
}
