// @a0n/gnosis/distributed-inference-host/src/style-transfer-mesh-gate.ts
/**
 * Per-pixel admission gate for classical style transfer.
 *
 * `effectiveGate = clamp01(alphaWave * topologyAlpha * depthConfidence * photometricVeto)`.
 * Missing optional maps default to all-ones (no extra veto).
 *
 * Six named admission channels (telemetry only; not Lean-derived constants):
 * see {@link STYLE_TRANSFER_MESH_SIX_CHANNEL_IDS}.
 */

import type {
  StyleTransferMeshAdmissionChannel,
  StyleTransferMeshFoldMetrics,
} from './style-transfer-types';

export const STYLE_TRANSFER_MESH_GATE_SCHEMA_VERSION = 1 as const;

/** Engineering channel ids for provenance / replay (documentation + telemetry). */
export const STYLE_TRANSFER_MESH_SIX_CHANNEL_IDS = [
  'edge_coherence_wave',
  'topology_sieve',
  'depth_confidence',
  'photometric_veto',
  'ensemble_disagreement',
  'carrier_posture',
] as const;

export type StyleTransferMeshSixChannelId = (typeof STYLE_TRANSFER_MESH_SIX_CHANNEL_IDS)[number];

export interface BuildEffectiveGateMapArgs {
  readonly pixelCount: number;
  readonly alphaWave: Float32Array;
  readonly topologyAlpha?: Float32Array | null;
  readonly depthConfidence?: Float32Array | null;
  readonly photometricVeto?: Float32Array | null;
}

/**
 * Builds the Effective Gate Map.
 */
export function buildEffectiveGateMap(args: BuildEffectiveGateMapArgs): Float32Array {
  const { pixelCount, alphaWave, topologyAlpha, depthConfidence, photometricVeto } = args;
  const out = new Float32Array(pixelCount);
  for (let i = 0; i < pixelCount; i++) {
    out[i] = effectiveGateAtPixel(
      alphaWave[i]!,
      topologyAlpha?.[i],
      depthConfidence?.[i],
      photometricVeto?.[i],
    );
  }
  return out;
}

/**
 * Handles the gnosis effective Gate At Pixel workflow.
 */
export function effectiveGateAtPixel(
  alphaWave: number,
  topologyAlpha?: number,
  depthConfidence?: number,
  photometricVeto?: number,
): number {
  let g = clamp01(alphaWave);
  if (topologyAlpha !== undefined) g *= clamp01(topologyAlpha);
  if (depthConfidence !== undefined) g *= clamp01(depthConfidence);
  if (photometricVeto !== undefined) g *= clamp01(photometricVeto);
  return clamp01(g);
}

export interface MeshGateDiagnostics {
  readonly meanAlphaWave: number;
  readonly meanTopologyAlpha: number;
  readonly meanDepthConfidence: number;
  readonly meanPhotometricVeto: number;
  readonly meanEffectiveGate: number;
  /** Higher = more admission blocked (1 - mean gate). */
  readonly gateViolationScore: number;
  /** Std dev of per-pixel effective gate (proxy for local ensemble spread). */
  readonly effectiveGateStd: number;
  /** Product of mean topology × depth × photometric admission (carrier coherence). */
  readonly carrierCoherenceMean: number;
}

/**
 * Computes the Mesh Gate Diagnostics.
 */
export function computeMeshGateDiagnostics(
  alphaWave: Float32Array,
  effectiveGate: Float32Array,
  topologyAlpha: Float32Array | null | undefined,
  depthConfidence: Float32Array | null | undefined,
  photometricVeto: Float32Array | null | undefined,
): MeshGateDiagnostics {
  const n = effectiveGate.length;
  let sumWave = 0;
  let sumTopo = 0;
  let sumDepth = 0;
  let sumVeto = 0;
  let sumGate = 0;
  const hasTopo = topologyAlpha !== null && topologyAlpha !== undefined && topologyAlpha.length === n;
  const hasDepth = depthConfidence !== null && depthConfidence !== undefined && depthConfidence.length === n;
  const hasVeto = photometricVeto !== null && photometricVeto !== undefined && photometricVeto.length === n;
  for (let i = 0; i < n; i++) {
    sumWave += alphaWave[i]!;
    sumGate += effectiveGate[i]!;
    if (hasTopo) sumTopo += topologyAlpha![i]!;
    if (hasDepth) sumDepth += depthConfidence![i]!;
    if (hasVeto) sumVeto += photometricVeto![i]!;
  }
  const inv = 1 / Math.max(1, n);
  const meanEffectiveGate = sumGate * inv;
  let sumSq = 0;
  for (let i = 0; i < n; i++) {
    const d = effectiveGate[i]! - meanEffectiveGate;
    sumSq += d * d;
  }
  const effectiveGateStd = Math.sqrt(sumSq * inv);
  const mTopo = hasTopo ? sumTopo * inv : 1;
  const mDepth = hasDepth ? sumDepth * inv : 1;
  const mVeto = hasVeto ? sumVeto * inv : 1;
  return {
    meanAlphaWave: sumWave * inv,
    meanTopologyAlpha: mTopo,
    meanDepthConfidence: mDepth,
    meanPhotometricVeto: mVeto,
    meanEffectiveGate,
    gateViolationScore: clamp01(1 - meanEffectiveGate),
    effectiveGateStd,
    carrierCoherenceMean: clamp01(mTopo * mDepth * mVeto),
  };
}

/**
 * Builds six-channel provenance rows. Channels without active buffers use
 * neutral admissibleWeight 1 and meanSignal from diagnostics when available.
 */
export function buildSixChannelAdmissionRows(d: MeshGateDiagnostics): StyleTransferMeshAdmissionChannel[] {
  return [
    {
      channelId: STYLE_TRANSFER_MESH_SIX_CHANNEL_IDS[0],
      meanSignal: d.meanAlphaWave,
      admissibleWeight: d.meanAlphaWave,
    },
    {
      channelId: STYLE_TRANSFER_MESH_SIX_CHANNEL_IDS[1],
      meanSignal: d.meanTopologyAlpha,
      admissibleWeight: d.meanTopologyAlpha,
    },
    {
      channelId: STYLE_TRANSFER_MESH_SIX_CHANNEL_IDS[2],
      meanSignal: d.meanDepthConfidence,
      admissibleWeight: d.meanDepthConfidence,
    },
    {
      channelId: STYLE_TRANSFER_MESH_SIX_CHANNEL_IDS[3],
      meanSignal: d.meanPhotometricVeto,
      admissibleWeight: d.meanPhotometricVeto,
    },
    {
      channelId: STYLE_TRANSFER_MESH_SIX_CHANNEL_IDS[4],
      meanSignal: d.effectiveGateStd,
      admissibleWeight: clamp01(1 - d.effectiveGateStd),
    },
    {
      channelId: STYLE_TRANSFER_MESH_SIX_CHANNEL_IDS[5],
      meanSignal: d.carrierCoherenceMean,
      admissibleWeight: d.carrierCoherenceMean,
    },
  ];
}

function clamp01(v: number): number {
  if (v <= 0) return 0;
  if (v >= 1) return 1;
  return v;
}

/**
 * Fold ranking beyond raw styleDistance: penalize low topology / depth coherence.
 * Lower composite score is better (same direction as styleDistance).
 */
export function compositeStyleTransferFoldScore(
  styleDistance: number,
  metrics: StyleTransferMeshFoldMetrics | undefined,
  weights: { topology?: number; depth?: number },
): number {
  if (metrics === undefined) return styleDistance;
  const wTopo = weights.topology ?? 0.15;
  const wDepth = weights.depth ?? 0.15;
  return (
    styleDistance +
    wTopo * (1 - clamp01(metrics.mean_topology_alpha)) +
    wDepth * (1 - clamp01(metrics.mean_depth_confidence))
  );
}
