// @a0n/gnosis/distributed-inference-host/src/style-transfer-mesh-depth-glue.ts
/**
 * Optional mesh depth confidence for style admission.
 *
 * When `build_mesh_depth_field` (Rust) is exposed to the host as WASM, wire
 * its `confidence` buffer into `StyleTransferAnyImageOptions.meshGate.depthConfidence`.
 * This module only validates buffers and emits JSON fragments for provenance.
 */

export const STYLE_TRANSFER_MESH_DEPTH_GLUE_SCHEMA_VERSION = 1 as const;

/** Validates optional per-pixel Float32 buffers (depth confidence, photometric veto, …). */
export function normalizePerPixelBuffer(
  buf: Float32Array | undefined,
  pixelCount: number,
): Float32Array | null {
  if (buf === undefined) return null;
  if (buf.length !== pixelCount) return null;
  return buf;
}

/**
 * Handles the gnosis mesh Depth Provenance Unused workflow.
 */
export function meshDepthProvenanceUnused(): Readonly<Record<string, unknown>> {
  return {
    schema_version: STYLE_TRANSFER_MESH_DEPTH_GLUE_SCHEMA_VERSION,
    build_mesh_depth_field: 'not_invoked',
    note: 'Pass meshGate.depthConfidence (length width*height, values in [0,1]) when WASM/native exposes build_mesh_depth_field.',
  };
}

/**
 * Handles the gnosis mesh Depth Provenance From Caller Buffer workflow.
 */
export function meshDepthProvenanceFromCallerBuffer(meanConfidence: number): Readonly<Record<string, unknown>> {
  return {
    schema_version: STYLE_TRANSFER_MESH_DEPTH_GLUE_SCHEMA_VERSION,
    source: 'caller_buffer',
    mean_confidence: meanConfidence,
  };
}
