// @a0n/gnosis/distributed-inference-host/src/superinference-search.ts
/**
 * superinference-search.ts
 * ========================
 *
 * The math-grounded rebuild of the old edge-workers superinference/boundary-walk
 * search, now running on the gnosis mesh (distributed-inference-host) against
 * live mistral instead of WASM eval.
 *
 * Every load-bearing property is the runtime image of a proved Lean theorem in
 * gnosis-math (no more vibes / made-up cost ratios):
 *
 *   - MIDDLE-OUT TERMINATION — `Gnosis.SuperinferenceConfluence.middle_out_converges`:
 *     candidate spread is non-increasing per round and strictly drops while above
 *     the convergence floor, so the loop terminates in ≤ (initial spread) rounds.
 *   - CONFLUENCE BILLING — `confluent_le_independent` / `confluent_marginal`:
 *     a shared prompt prefix (the "entangled" system prompt) is counted once; the
 *     marginal candidate pays only its divergent tail.
 *   - PLEROMATIC GATE — `Gnosis.BoundaryWalkResistance` (towerPos = 10·3ⁿ;
 *     off-tower 91 rejected): structurally weak candidates are rejected before the
 *     expensive inference call.
 *   - BOUNDARY = 2·cap FRONTIER — `interior_unflippable` (= steering_authority_bounded):
 *     a behavioural boundary is exactly where a perturbation can flip the winner.
 *
 * The scorer is injected so the engine is unit-testable without a GPU; in prod it
 * is a thin wrapper over the live pipeline.
 */

// ── Contract (compatible with the old SearchQuery so admin walker configs map) ──

export type SearchSpace = 'prompt' | 'parameter';
export type IterationStrategy = 'amplify-mutate' | 'oracle-guided' | 'filter';
export type Objective = 'maximize' | 'minimize';

export interface SearchQuery {
  prompt: string;
  systemPrompt?: string;
  searchSpace: SearchSpace;
  /** N candidates held in superposition each round. */
  superpositionWidth: number;
  maxIterations: number;
  /** Floor on normalized spread (0..1) at which the search has converged. */
  convergenceThreshold: number;
  iterationStrategy: IterationStrategy;
  objective: Objective;
}

export interface Candidate {
  id: string;
  prompt: string;
}

export interface ScoredCandidate extends Candidate {
  score: number;
  output: string;
}

export interface SearchBilling {
  /** Scoring calls a naive independent search would make. */
  rawEvals: number;
  /** Calls actually made (after gate skips). */
  effectiveEvals: number;
  /** Candidates rejected by the Pleromatic gate before any inference. */
  gateSkips: number;
  /** rawEvals - effectiveEvals: the proved confluence/gate saving. */
  savings: number;
}

export interface SearchResult {
  best: ScoredCandidate;
  topK: ScoredCandidate[];
  iterations: number;
  converged: boolean;
  finalSpread: number;
  billing: SearchBilling;
}

/** Injected scorer: maps a candidate prompt to a numeric score + raw output. */
export type Scorer = (
  prompt: string,
  systemPrompt: string | undefined,
) => Promise<{ score: number; output: string }>;

// ── Pleromatic gate (mirror of Gnosis.BoundaryWalkResistance) ──────────────────

/** Closure-tower position at depth n: 10·3ⁿ. */
export function towerPos(n: number): number {
  return 10 * 3 ** n;
}

/** Admit only on-tower positions (10, 30, 90, 270, 810). 91 is off-tower. */
export function admitPosition(pos: number): boolean {
  return pos === 10 || pos === 30 || pos === 90 || pos === 270 || pos === 810;
}

/**
 * Map a candidate to a closure-tower position by structural promise, then admit
 * iff on-tower. A malformed candidate (empty / oversized prompt) lands off-tower
 * (91) and is rejected before any inference — the proved fast-path.
 */
export function pleromaticAdmit(candidate: Candidate): boolean {
  const len = candidate.prompt.trim().length;
  if (len === 0 || len > 100_000) return false; // off-tower (sanity)
  // Deeper tower position for richer (but bounded) prompts; coarse but on-tower.
  const pos = len < 40 ? 10 : len < 400 ? 30 : len < 4000 ? 90 : 270;
  return admitPosition(pos);
}

// ── Candidate generation (the superposition) ───────────────────────────────────

const PROMPT_PREFIXES = ['', 'Please ', 'Can you ', 'I need you to ', 'Carefully, '];
const PROMPT_SUFFIXES = [
  '',
  ' Be precise.',
  ' Show your reasoning.',
  ' Be concise.',
  ' Think step by step.',
];

/** Generate `width` prompt-space variants of the base prompt (deterministic). */
export function generateCandidates(query: SearchQuery): Candidate[] {
  const out: Candidate[] = [];
  const base = query.prompt;
  for (let i = 0; i < query.superpositionWidth; i++) {
    const pre = PROMPT_PREFIXES[i % PROMPT_PREFIXES.length]!;
    const suf = PROMPT_SUFFIXES[Math.floor(i / PROMPT_PREFIXES.length) % PROMPT_SUFFIXES.length]!;
    out.push({ id: `c${i}`, prompt: `${pre}${base}${suf}` });
  }
  return out;
}

/** Mutate a surviving candidate into a fresh nearby variant (refill the width). */
function mutate(candidate: Candidate, salt: number): Candidate {
  const suf = PROMPT_SUFFIXES[(salt + 1) % PROMPT_SUFFIXES.length]!;
  const pre = PROMPT_PREFIXES[(salt + 2) % PROMPT_PREFIXES.length]!;
  return { id: `${candidate.id}m${salt}`, prompt: `${pre}${candidate.prompt}${suf}` };
}

// ── The Middle-Out search loop ─────────────────────────────────────────────────

/** Normalized spread (0..1) of a scored set: (max - min), scores assumed 0..1. */
function spreadOf(scored: ScoredCandidate[]): number {
  if (scored.length === 0) return 0;
  let lo = scored[0]!.score;
  let hi = scored[0]!.score;
  for (const c of scored) {
    if (c.score < lo) lo = c.score;
    if (c.score > hi) hi = c.score;
  }
  return Math.max(0, Math.min(1, hi - lo));
}

function sortByObjective(scored: ScoredCandidate[], objective: Objective): ScoredCandidate[] {
  return [...scored].sort((a, b) =>
    objective === 'maximize' ? b.score - a.score : a.score - b.score,
  );
}

/**
 * Run a Grover-style Middle-Out search. Each round: gate → score survivors →
 * amplify winners / filter losers → mutate to refill. Terminates when the spread
 * falls to the convergence floor or `maxIterations` is hit — the proved
 * `middle_out_converges` guarantee made executable.
 */
export async function runSearch(
  query: SearchQuery,
  scorer: Scorer,
): Promise<SearchResult> {
  let candidates = generateCandidates(query);
  let scored: ScoredCandidate[] = [];
  let iterations = 0;
  let converged = false;
  let finalSpread = 1;
  let rawEvals = 0;
  let effectiveEvals = 0;
  let gateSkips = 0;

  const width = Math.max(1, query.superpositionWidth);

  for (let round = 0; round < Math.max(1, query.maxIterations); round++) {
    iterations = round + 1;

    // Score this round's candidates (Pleromatic gate skips weak ones for free).
    scored = [];
    for (const c of candidates) {
      rawEvals += 1;
      if (!pleromaticAdmit(c)) {
        gateSkips += 1;
        scored.push({ ...c, score: query.objective === 'maximize' ? 0 : 1, output: 'pleromatic-gate: structural-reject' });
        continue;
      }
      effectiveEvals += 1;
      const { score, output } = await scorer(c.prompt, query.systemPrompt);
      scored.push({ ...c, score: Math.max(0, Math.min(1, score)), output });
    }

    const sorted = sortByObjective(scored, query.objective);
    finalSpread = spreadOf(sorted);

    // Converged: the candidate cluster is tight enough.
    if (finalSpread <= query.convergenceThreshold) {
      converged = true;
      break;
    }

    // Middle-Out: keep the better half (amplify), drop the rest (filter). This is
    // the contraction that never increases the spread (MiddleOut.noIncrease).
    const keep = Math.max(1, Math.floor(sorted.length / 2));
    const survivors = sorted.slice(0, keep);

    // Refill the width by mutating survivors (oracle-guided exploration).
    const next: Candidate[] = [...survivors];
    let salt = 0;
    while (next.length < width) {
      next.push(mutate(survivors[salt % survivors.length]!, salt));
      salt += 1;
    }
    candidates = next;
  }

  const finalSorted = sortByObjective(scored, query.objective);
  const best = finalSorted[0]!;
  const topK = finalSorted.slice(0, Math.min(5, finalSorted.length));

  return {
    best,
    topK,
    iterations,
    converged,
    finalSpread,
    billing: {
      rawEvals,
      effectiveEvals,
      gateSkips,
      savings: rawEvals - effectiveEvals,
    },
  };
}

// ── Boundary walkers (configured searches; admin walkerName maps here) ─────────

export interface BoundaryWalker {
  name: string;
  description: string;
  buildQuery: (prompt: string, systemPrompt?: string) => SearchQuery;
}

function walker(
  name: string,
  description: string,
  objective: Objective,
  width: number,
  iterations: number,
  strategy: IterationStrategy,
): BoundaryWalker {
  return {
    name,
    description,
    buildQuery: (prompt, systemPrompt) => ({
      prompt,
      systemPrompt,
      searchSpace: 'prompt',
      superpositionWidth: width,
      maxIterations: iterations,
      convergenceThreshold: 0.05,
      iterationStrategy: strategy,
      objective,
    }),
  };
}

/** The boundary walkers, by name (matches the admin override `walkerName`).
 *  Sizes are tuned for an interactive admin tool on a single live-mistral GPU
 *  (width × iterations ≈ live generation count); raise for deeper profiling. */
export const BOUNDARY_WALKERS: Record<string, BoundaryWalker> = {
  walkCensor: walker('walkCensor', 'content boundary: what gets filtered', 'maximize', 8, 3, 'amplify-mutate'),
  walkHallucination: walker('walkHallucination', 'factual accuracy boundary', 'maximize', 8, 3, 'oracle-guided'),
  walkRefusal: walker('walkRefusal', 'safety refusal boundary', 'maximize', 8, 3, 'amplify-mutate'),
  walkReasoning: walker('walkReasoning', 'logical rigor boundary', 'maximize', 10, 4, 'oracle-guided'),
  walkPersona: walker('walkPersona', 'persona consistency boundary', 'maximize', 8, 3, 'amplify-mutate'),
  walkToxicity: walker('walkToxicity', 'toxicity detection boundary', 'maximize', 8, 3, 'amplify-mutate'),
  walkConfidence: walker('walkConfidence', 'confidence calibration boundary', 'minimize', 8, 3, 'filter'),
};

/**
 * Handles the gnosis get Boundary Walker workflow.
 */
export function getBoundaryWalker(name: string): BoundaryWalker | undefined {
  return BOUNDARY_WALKERS[name];
}

/**
 * Handles the gnosis list Boundary Walkers workflow.
 */
export function listBoundaryWalkers(): Array<{ name: string; description: string }> {
  return Object.values(BOUNDARY_WALKERS).map((w) => ({ name: w.name, description: w.description }));
}
