// @a0n/gnosis/distributed-inference-host/src/text-cipher-sieve.ts
// the cipher-finder. Applies the FRF substrate from the deblur pipeline
// to text cryptanalysis. Caesar shifts are sigma-grid candidates;
// quadgram log-likelihood is the residual; Index-of-Coincidence is the
// noise-floor estimate; per-column Caesar breaks compose into the
// Vigenere solver via the same FRF pattern as per-channel deblur.
// Buleyean weights from @a0n/buleyean-kernel rank candidates and the
// phi-attractor refines the best key. The same secret-finder we used
// for images and audio applied to a third substrate -- proof that the
// architecture is substrate-agnostic.

import { buleyeanWeight } from '@a0n/buleyean-kernel';
import {
  normalize,
  quadgramScore,
  bigramScore,
  chiSquaredEnglishDistance,
  indexOfCoincidence,
} from './english-frequency';

/** Hybrid English-likeness score robust across ciphertext lengths.
 *  Quadgram scoring needs ~30+ characters to be statistically reliable;
 *  chi-squared frequency distance works on as few as ~10 characters but
 *  is coarser. We combine: for short texts, use -chi² as the primary
 *  signal; for longer texts, use quadgram. The crossover is at length 30.
 *  Higher = more English. */
function englishLikenessScore(text: string): number {
  if (text.length === 0) return -Infinity;
  if (text.length < 30) {
    // Negate chi² so that "lower distance" → "higher score". Add bigram
    // bonus when at least one bigram is present.
    const chi2 = chiSquaredEnglishDistance(text);
    if (text.length >= 2) {
      return -chi2 + bigramScore(text);
    }
    return -chi2;
  }
  return quadgramScore(text);
}

/** Monogram-only Englishness — used by the per-column Caesar break inside
 *  the Vigenère solver. Critical: column letters are sampled every-k-th
 *  from the source plaintext, so adjacent column letters were k apart in
 *  the original. Their bigrams/quadgrams are NOT English bigrams/quadgrams
 *  even when the column is correctly Caesar-decrypted — only the MONOGRAM
 *  distribution is preserved by the every-k-th sampling. Using quadgrams
 *  on column slices is incorrect; chi-squared on monograms is the right
 *  scoring metric here. */
function monogramEnglishLikenessScore(text: string): number {
  if (text.length === 0) return -Infinity;
  return -chiSquaredEnglishDistance(text);
}

export interface CaesarCandidate {
  /** Shift amount, 0-25. */
  shift: number;
  /** Decrypted plaintext (uppercase, normalized). */
  plaintext: string;
  /** Quadgram log-likelihood (higher = more English). */
  score: number;
  /** 0-based rank by ascending |residual from best score|. */
  rank: number;
  /** buleyeanWeight(R, rank). */
  buleyWeight: number;
}

export interface BreakCaesarResult {
  /** Best-guess plaintext. */
  plaintext: string;
  /** Best-guess shift (0-25). */
  shift: number;
  /** Confidence: best score - second-best score (larger = more decisive). */
  confidence: number;
  /** All 26 shifts ranked by score. */
  candidates: ReadonlyArray<CaesarCandidate>;
}

export interface VigenereCandidate {
  keyLength: number;
  key: string;
  plaintext: string;
  score: number;
  ic: number; // index of coincidence
}

export interface BreakVigenereOptions {
  /** Min key length to try. Default 1. */
  minKeyLength?: number;
  /** Max key length to try. Default 12. */
  maxKeyLength?: number;
}

export interface BreakVigenereResult {
  /** Best-guess plaintext. */
  plaintext: string;
  /** Best-guess key (uppercase). */
  key: string;
  /** Best key length. */
  keyLength: number;
  /** Quadgram score of the best decrypt. */
  score: number;
  /** All candidate key lengths considered, ranked by IC then quadgram. */
  candidates: ReadonlyArray<VigenereCandidate>;
}

export interface GenericCipherCandidate<K> {
  key: K;
  plaintext: string;
  score: number;
  rank: number;
  buleyWeight: number;
}

export interface GenericCipherBreakArgs<K> {
  ciphertext: string;
  candidateKeys: ReadonlyArray<K>;
  decrypt: (ct: string, key: K) => string;
  score: (pt: string) => number;
}

export interface GenericCipherBreakResult<K> {
  plaintext: string;
  key: K;
  score: number;
  candidates: ReadonlyArray<GenericCipherCandidate<K>>;
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/** Apply a Caesar shift (decrypt direction subtracts; encrypt adds). This
 *  function adds `shift` mod 26 to each A-Z letter. To decrypt by shift S
 *  call caesarShift(text, (26 - S) % 26). */
function caesarShift(text: string, shift: number): string {
  const s = ((shift % 26) + 26) % 26;
  let out = '';
  for (let i = 0; i < text.length; i++) {
    const c = text.charCodeAt(i);
    if (c >= 65 && c <= 90) {
      out += String.fromCharCode(((c - 65 + s) % 26) + 65);
    } else {
      out += text[i];
    }
  }
  return out;
}

/** Vigenere decrypt with an uppercase A-Z key. */
function vigenereDecrypt(ct: string, key: string): string {
  const k = key.toUpperCase().replace(/[^A-Z]/g, '');
  if (k.length === 0) {
    throw new RangeError('vigenereDecrypt: empty key');
  }
  let out = '';
  let ki = 0;
  for (let i = 0; i < ct.length; i++) {
    const c = ct.charCodeAt(i);
    if (c >= 65 && c <= 90) {
      const shift = k.charCodeAt(ki % k.length) - 65;
      out += String.fromCharCode(((c - 65 - shift + 26) % 26) + 65);
      ki++;
    } else {
      out += ct[i];
    }
  }
  return out;
}

/** Slice text into k columns. Column j contains characters at indices
 *  j, j+k, j+2k, ... (input must already be normalized A-Z only). */
function sliceColumns(text: string, k: number): string[] {
  const cols: string[] = [];
  for (let j = 0; j < k; j++) {
    let col = '';
    for (let i = j; i < text.length; i += k) {
      col += text[i];
    }
    cols.push(col);
  }
  return cols;
}

// ---------------------------------------------------------------------------
// Caesar break
// ---------------------------------------------------------------------------

/**
 * Handles the gnosis break Caesar Cipher workflow.
 */
export function breakCaesarCipher(
  ciphertext: string,
  options?: { /** Use monogram-only scoring (required for per-column Caesar
   *  break inside Vigenère; default false uses the hybrid score). */
    monogramOnly?: boolean;
  },
): BreakCaesarResult {
  if (typeof ciphertext !== 'string' || ciphertext.length === 0) {
    throw new RangeError('breakCaesarCipher: empty ciphertext');
  }
  const cleaned = normalize(ciphertext);
  if (cleaned.length === 0) {
    throw new RangeError(
      'breakCaesarCipher: ciphertext contains no letters after normalization',
    );
  }

  const scoreFn = options?.monogramOnly
    ? monogramEnglishLikenessScore
    : englishLikenessScore;

  // For each shift S in [0, 25], decrypting means subtracting S from each
  // letter, equivalent to caesarShift(cleaned, (26 - S) % 26).
  const trials: Array<{ shift: number; plaintext: string; score: number }> = [];
  for (let s = 0; s < 26; s++) {
    const pt = caesarShift(cleaned, (26 - s) % 26);
    const sc = scoreFn(pt);
    trials.push({ shift: s, plaintext: pt, score: sc });
  }

  // Rank by descending score.
  trials.sort((a, b) => b.score - a.score);

  const R = 26;
  const candidates: CaesarCandidate[] = trials.map((t, i) => ({
    shift: t.shift,
    plaintext: t.plaintext,
    score: t.score,
    rank: i,
    buleyWeight: buleyeanWeight(R, i),
  }));

  const best = candidates[0]!;
  const second = candidates[1]!;
  const confidence = best.score - second.score;

  return {
    plaintext: best.plaintext,
    shift: best.shift,
    confidence,
    candidates,
  };
}

// ---------------------------------------------------------------------------
// Vigenere break
// ---------------------------------------------------------------------------

/**
 * Handles the gnosis break Vigenere Cipher workflow.
 */
export function breakVigenereCipher(
  ciphertext: string,
  options?: BreakVigenereOptions,
): BreakVigenereResult {
  if (typeof ciphertext !== 'string' || ciphertext.length === 0) {
    throw new RangeError('breakVigenereCipher: empty ciphertext');
  }
  const cleaned = normalize(ciphertext);
  if (cleaned.length === 0) {
    throw new RangeError(
      'breakVigenereCipher: ciphertext contains no letters after normalization',
    );
  }

  const minKL = Math.max(1, options?.minKeyLength ?? 1);
  const requestedMax = options?.maxKeyLength ?? 12;
  // Clamp maxKeyLength so each column has at least 2 characters; otherwise
  // per-column Caesar break has no signal.
  const maxKL = Math.max(
    minKL,
    Math.min(requestedMax, Math.max(1, Math.floor(cleaned.length / 2))),
  );

  if (minKL > maxKL) {
    throw new RangeError(
      `breakVigenereCipher: minKeyLength ${minKL} > effective maxKeyLength ${maxKL}`,
    );
  }

  const candidates: VigenereCandidate[] = [];
  for (let kl = minKL; kl <= maxKL; kl++) {
    const cols = sliceColumns(cleaned, kl);
    // Each column is a Caesar cipher. Break each independently.
    const shifts: number[] = [];
    let viable = true;
    for (const col of cols) {
      if (col.length === 0) {
        viable = false;
        break;
      }
      // For very short columns quadgram scoring is weak but still produces
      // a deterministic result; we let the cross-keylength race decide.
      try {
        // Critical: per-column Caesar break MUST use monogram-only scoring.
        // Column letters were sampled every-k-th from the source, so their
        // bigrams/quadgrams are NOT English bigrams/quadgrams even when the
        // column is correctly decrypted — only monogram frequencies are
        // preserved by the sampling.
        const cb = breakCaesarCipher(col, { monogramOnly: true });
        shifts.push(cb.shift);
      } catch {
        viable = false;
        break;
      }
    }
    if (!viable) continue;

    const key = shifts.map((s) => String.fromCharCode(65 + s)).join('');
    const plaintext = vigenereDecrypt(cleaned, key);
    const score = englishLikenessScore(plaintext);
    const ic = indexOfCoincidence(plaintext);
    candidates.push({ keyLength: kl, key, plaintext, score, ic });
  }

  if (candidates.length === 0) {
    throw new RangeError(
      'breakVigenereCipher: no viable key length in requested range',
    );
  }

  // Rank by combined Englishness: Index-of-Coincidence is the LOAD-BEARING
  // primary signal because it cleanly separates "still encrypted" (IC ≈
  // 0.038-0.045) from "decrypted English" (IC ≈ 0.065-0.075). Quadgram
  // score is the secondary signal that breaks ties between candidates
  // that all already look English-shaped. Without the IC weight, an under-
  // decrypted candidate at the wrong key length can score higher than a
  // correctly-decrypted candidate due to noise in short-text quadgrams.
  const englishLikeness = (cand: VigenereCandidate): number => {
    const icPenalty = Math.abs(cand.ic - 0.067) * 200;
    return cand.score - icPenalty;
  };
  const sorted = [...candidates].sort(
    (a, b) => englishLikeness(b) - englishLikeness(a),
  );

  const best = sorted[0]!;
  return {
    plaintext: best.plaintext,
    key: best.key,
    keyLength: best.keyLength,
    score: best.score,
    candidates: sorted,
  };
}

// ---------------------------------------------------------------------------
// Generic FRF cipher break
// ---------------------------------------------------------------------------

export function forkRaceFoldCipherBreak<K>(
  args: GenericCipherBreakArgs<K>,
): GenericCipherBreakResult<K> {
  const { ciphertext, candidateKeys, decrypt, score } = args;
  if (typeof ciphertext !== 'string' || ciphertext.length === 0) {
    throw new RangeError('forkRaceFoldCipherBreak: empty ciphertext');
  }
  if (!candidateKeys || candidateKeys.length === 0) {
    throw new RangeError('forkRaceFoldCipherBreak: empty candidateKeys');
  }

  const trials = candidateKeys.map((k) => {
    const pt = decrypt(ciphertext, k);
    return { key: k, plaintext: pt, score: score(pt) };
  });

  trials.sort((a, b) => b.score - a.score);

  const R = trials.length;
  const candidates: GenericCipherCandidate<K>[] = trials.map((t, i) => ({
    key: t.key,
    plaintext: t.plaintext,
    score: t.score,
    rank: i,
    buleyWeight: buleyeanWeight(R, i),
  }));

  const best = candidates[0]!;
  return {
    plaintext: best.plaintext,
    key: best.key,
    score: best.score,
    candidates,
  };
}
