// @a0n/gnosis/distributed-inference-host/src/thoth-oath-tools.ts
import type {
  AgenticMcpClient,
  AgenticToolDefinition,
  ToolExecutionResult,
} from './agentic-chat';
import {
  AEON,
  PHI_APERIODIC_ROTATION_STEP,
  phaseFromPhiTheta,
  phaseOfNatTick,
} from '../../src/gnosis-time/index';

export const THOTH_OATH_PROPOSE_TOOL = 'thoth/oath.propose' as const;
export const THOTH_OATH_CLEAR_TOOL = 'thoth/oath.clear' as const;
export const THOTH_OATH_AUDIT_TOOL = 'thoth/oath.audit' as const;

export interface ThothOathRecord {
  id: string;
  invariant: string;
  scope: string;
  consent: string;
  bindingStrength: 'operator-clearable' | 'mast-bound';
  status: 'proposed' | 'active' | 'cleared';
  source: string;
  sourceId?: string;
  theoremRef: string;
  createdAt: number;
  schedule: ThothOathSchedule;
  durationMs: number;
  expiresAt: number;
  clearedAt?: number;
  clearReason?: string;
}

export type ThothOathTelemetryEventKind =
  | 'oath.proposed'
  | 'oath.cleared'
  | 'oath.expired'
  | 'oath.audit';

export interface ThothOathTelemetryEvent {
  kind: ThothOathTelemetryEventKind;
  oathId?: string;
  invariant?: string;
  schedule?: ThothOathSchedule;
  effectiveStatus?: ThothOathRecord['status'] | 'expired';
  activeOathCount: number;
  theoremRef: string;
  nonAuthoritative: true;
  observedAt: number;
}

export type ThothOathScheduleMode = 'periodic' | 'aperiodic';

export interface ThothOathSchedule {
  mode: ThothOathScheduleMode;
  startTick: number;
  startPhase: number;
  durationTicks: number;
  expiresAtTick: number;
  expiresAtPhase: number;
  aeon: typeof AEON;
  aperiodicTheta?: number;
  aperiodicStep?: typeof PHI_APERIODIC_ROTATION_STEP;
}

export interface ThothOathToolOptions {
  now?: () => number;
  randomId?: (prefix: string) => string;
  ledgerStore?: ThothOathLedgerStore;
}

export interface ThothOathLedgerStore {
  load(): Promise<readonly ThothOathRecord[]>;
  save(records: readonly ThothOathRecord[]): Promise<void>;
}

const OATH_THEOREM_REF =
  'Gnosis.Witnesses.Interfaith.OathStabilityGateWitness.oath_stability_gate_witness';

export const THOTH_OATH_TOOL_DEFINITIONS: readonly AgenticToolDefinition[] = [
  {
    type: 'function',
    function: {
      name: THOTH_OATH_PROPOSE_TOOL,
      description:
        'Propose or program a bounded Thoth oath. Oaths bind a declared invariant for stability; they do not grant authority.',
      parameters: {
        type: 'object',
        properties: {
          invariant: {
            type: 'string',
            description: 'The invariant Thoth is committing to hold stable.',
          },
          scope: {
            type: 'string',
            description: 'Where the oath applies: prompt, nudge, action, or session.',
          },
          consent: {
            type: 'string',
            description:
              'Explicit consent or operator basis for the oath. Required to prevent capture.',
          },
          durationMs: {
            type: 'number',
            description:
              'Legacy wall-clock fallback. Prefer durationTicks on GnosisTime.',
          },
          durationTicks: {
            type: 'number',
            description:
              'How many GnosisTime ticks the oath remains active. Defaults to one aeon.',
          },
          startTick: {
            type: 'number',
            description:
              'Nonnegative GnosisClock/GnosisTime tick where the oath begins.',
          },
          scheduleMode: {
            type: 'string',
            description:
              'periodic for ordinary GnosisTime ticks, aperiodic for phi-gated scheduling.',
          },
          aperiodicTheta: {
            type: 'number',
            description:
              'Optional theta in [0,1) for aperiodic phi scheduling.',
          },
          bindingStrength: {
            type: 'string',
            description:
              'operator-clearable for repairable oaths, mast-bound for user-confirmed commitments that Thoth cannot clear.',
          },
          source: {
            type: 'string',
            description:
              'Optional external source namespace for audit provenance, for example shared-ui.microcontract.double-confirmation.',
          },
          sourceId: {
            type: 'string',
            description:
              'Optional external source record id, such as the originating microcontract proposal id.',
          },
          theoremRef: {
            type: 'string',
            description:
              'Optional theorem or witness reference backing the oath source.',
          },
        },
        required: ['invariant'],
      },
    },
    activation: {
      patterns: [
        '\\b(make|program|create|set|bind|take)\\b[\\s\\S]{0,80}\\boath\\b',
        '\\boath\\b[\\s\\S]{0,80}\\b(stabilize|commit|bind|program)\\b',
      ],
      keywords: ['oath', 'stability', 'commitment', 'invariant'],
      threshold: 0.68,
      argExtractors: {
        invariant: 'oath(?:\\s+to|\\s+that)?\\s+([^.;\\n]+)',
      },
    },
    source: 'provided',
    metadata: {
      thothSelfManagement: true,
      theoremRef: OATH_THEOREM_REF,
    },
  },
  {
    type: 'function',
    function: {
      name: THOTH_OATH_CLEAR_TOOL,
      description:
        'Clear a Thoth oath by id or by invariant text. Clearing preserves an audit trace instead of pretending the oath never existed.',
      parameters: {
        type: 'object',
        properties: {
          id: { type: 'string', description: 'Oath id to clear.' },
          invariant: {
            type: 'string',
            description: 'Invariant text to match when no id is available.',
          },
          reason: {
            type: 'string',
            description: 'Why the oath is being cleared or repaired.',
          },
        },
      },
    },
    activation: {
      patterns: [
        '\\b(clear|release|drop|retire|repair|void)\\b[\\s\\S]{0,80}\\boath\\b',
        '\\boath\\b[\\s\\S]{0,80}\\b(clear|release|drop|retire|repair|void)\\b',
      ],
      keywords: ['clear oath', 'release oath', 'repair oath'],
      threshold: 0.68,
    },
    source: 'provided',
    metadata: {
      thothSelfManagement: true,
      theoremRef: OATH_THEOREM_REF,
    },
  },
  {
    type: 'function',
    function: {
      name: THOTH_OATH_AUDIT_TOOL,
      description:
        'Audit active and cleared Thoth oaths with source-backed non-authority boundaries.',
      parameters: {
        type: 'object',
        properties: {
          status: {
            type: 'string',
            description: 'Filter by proposed, active, cleared, or all.',
          },
        },
      },
    },
    activation: {
      patterns: [
        '\\b(audit|list|show|inspect)\\b[\\s\\S]{0,80}\\boaths?\\b',
        '\\boaths?\\b[\\s\\S]{0,80}\\b(audit|list|show|inspect)\\b',
      ],
      keywords: ['audit oath', 'oath ledger', 'active oath'],
      threshold: 0.65,
    },
    source: 'provided',
    metadata: {
      thothSelfManagement: true,
      theoremRef: OATH_THEOREM_REF,
    },
  },
];

export class ThothOathSelfManagementMcpClient implements AgenticMcpClient {
  readonly sessionId = 'thoth-oath-local';
  private readonly oathLedger: ThothOathRecord[] = [];
  private readonly now: () => number;
  private readonly randomId: (prefix: string) => string;
  private readonly ledgerStore?: ThothOathLedgerStore;
  private hydrated = false;

  constructor(options: ThothOathToolOptions = {}) {
    this.now = options.now ?? (() => Date.now());
    this.randomId =
      options.randomId ??
      ((prefix) => `${prefix}_${Math.random().toString(36).slice(2, 10)}`);
    this.ledgerStore = options.ledgerStore;
  }

  async listTools(): Promise<AgenticToolDefinition[]> {
    await this.hydrateLedger();
    return [...THOTH_OATH_TOOL_DEFINITIONS];
  }

  async preflight() {
    return {
      tools: await this.listTools(),
      sessionId: this.sessionId,
      durationMs: 0,
      cached: true,
    };
  }

  async callTool(
    name: string,
    args: Record<string, unknown>,
    options?: { toolCallId?: string },
  ): Promise<ToolExecutionResult> {
    await this.hydrateLedger();
    const startedAt = this.now();
    const toolCallId = options?.toolCallId ?? this.randomId('call');
    try {
      if (name === THOTH_OATH_PROPOSE_TOOL) {
        const proposed = this.propose(args);
        await this.persistLedger();
        return this.result(toolCallId, name, args, proposed, startedAt);
      }
      if (name === THOTH_OATH_CLEAR_TOOL) {
        const cleared = this.clear(args);
        await this.persistLedger();
        return this.result(toolCallId, name, args, cleared, startedAt);
      }
      if (name === THOTH_OATH_AUDIT_TOOL) {
        return this.result(toolCallId, name, args, this.audit(args), startedAt);
      }
    } catch (error) {
      return {
        toolCallId,
        toolName: name,
        args,
        result: null,
        success: false,
        durationMs: this.now() - startedAt,
        error: error instanceof Error ? error.message : String(error),
      };
    }
    return {
      toolCallId,
      toolName: name,
      args,
      result: null,
      success: false,
      durationMs: this.now() - startedAt,
      error: `Unknown Thoth oath tool: ${name}`,
    };
  }

  private propose(args: Record<string, unknown>) {
    const invariant =
      typeof args.invariant === 'string' && args.invariant.trim()
        ? args.invariant.trim()
        : 'hold stability without authority claim';
    const schedule = this.schedule(args);
    const durationMs = schedule.durationTicks * 60 * 1000;
    const createdAt = this.now();
    const record: ThothOathRecord = {
      id: this.randomId('oath'),
      invariant,
      scope: typeof args.scope === 'string' ? args.scope : 'session',
      consent:
        typeof args.consent === 'string' && args.consent.trim()
          ? args.consent.trim()
          : 'operator-requested',
      bindingStrength: this.bindingStrength(args.bindingStrength),
      status: 'active',
      source: this.trimmedString(args.source, 'thoth-self-management'),
      ...(typeof args.sourceId === 'string' && args.sourceId.trim()
        ? { sourceId: args.sourceId.trim() }
        : {}),
      theoremRef: this.trimmedString(args.theoremRef, OATH_THEOREM_REF),
      createdAt,
      schedule,
      durationMs,
      expiresAt: createdAt + durationMs,
    };
    this.oathLedger.push(record);
    return {
      action: 'oath.proposed',
      oath: record,
      telemetryEvent: this.telemetryEvent('oath.proposed', record),
      nonAuthoritative: true,
      captureBoundary: 'agent consent and operator origin remain separate',
      timerBoundary:
        'oath expires on the GnosisClock/GnosisTime tick boundary unless explicitly renewed',
    };
  }

  private schedule(args: Record<string, unknown>): ThothOathSchedule {
    const mode = this.scheduleMode(args.scheduleMode);
    const startTick = this.nonnegativeInteger(args.startTick, 'startTick', 0);
    const durationTicks =
      args.durationTicks !== undefined
        ? this.positiveInteger(args.durationTicks, 'durationTicks')
        : this.legacyDurationTicks(args);
    const expiresAtTick = startTick + durationTicks;
    const startPhase = phaseOfNatTick(startTick);
    const expiresAtPhase = phaseOfNatTick(expiresAtTick);
    if (mode === 'aperiodic') {
      const aperiodicTheta =
        typeof args.aperiodicTheta === 'number' && Number.isFinite(args.aperiodicTheta)
          ? args.aperiodicTheta
          : (startPhase / AEON + PHI_APERIODIC_ROTATION_STEP) % 1;
      return {
        mode,
        startTick,
        startPhase: phaseFromPhiTheta(aperiodicTheta),
        durationTicks,
        expiresAtTick,
        expiresAtPhase: phaseFromPhiTheta(
          aperiodicTheta + durationTicks * PHI_APERIODIC_ROTATION_STEP,
        ),
        aeon: AEON,
        aperiodicTheta,
        aperiodicStep: PHI_APERIODIC_ROTATION_STEP,
      };
    }
    return {
      mode,
      startTick,
      startPhase,
      durationTicks,
      expiresAtTick,
      expiresAtPhase,
      aeon: AEON,
    };
  }

  private scheduleMode(value: unknown): ThothOathScheduleMode {
    if (value === undefined || value === 'periodic') {
      return 'periodic';
    }
    if (value === 'aperiodic') {
      return 'aperiodic';
    }
    throw new RangeError(
      `Thoth oath scheduleMode expects periodic or aperiodic, got ${String(value)}`,
    );
  }

  private bindingStrength(value: unknown): ThothOathRecord['bindingStrength'] {
    if (value === undefined || value === 'operator-clearable') {
      return 'operator-clearable';
    }
    if (value === 'mast-bound') {
      return 'mast-bound';
    }
    throw new RangeError(
      `Thoth oath bindingStrength expects operator-clearable or mast-bound, got ${String(value)}`,
    );
  }

  private legacyDurationTicks(args: Record<string, unknown>): number {
    const legacyMs =
      typeof args.durationMs === 'number' && Number.isFinite(args.durationMs)
        ? args.durationMs
        : typeof args.ttlMs === 'number' && Number.isFinite(args.ttlMs)
          ? args.ttlMs
          : undefined;
    if (legacyMs === undefined) {
      return AEON;
    }
    const tickMs = 60 * 1000;
    if (!Number.isInteger(legacyMs) || legacyMs <= 0 || legacyMs % tickMs !== 0) {
      throw new RangeError(
        `Thoth oath durationMs must align to whole GnosisTime minute ticks, got ${String(legacyMs)}`,
      );
    }
    return this.positiveInteger(legacyMs / tickMs, 'durationTicks');
  }

  private positiveInteger(value: unknown, label: string): number {
    if (!Number.isInteger(value) || typeof value !== 'number' || value <= 0) {
      throw new RangeError(`Thoth oath ${label} expects positive integer, got ${String(value)}`);
    }
    return value;
  }

  private nonnegativeInteger(value: unknown, label: string, fallback: number): number {
    const effective = value ?? fallback;
    if (!Number.isInteger(effective) || typeof effective !== 'number' || effective < 0) {
      throw new RangeError(
        `Thoth oath ${label} expects nonnegative integer, got ${String(effective)}`,
      );
    }
    return effective;
  }

  private trimmedString(value: unknown, fallback: string): string {
    return typeof value === 'string' && value.trim() ? value.trim() : fallback;
  }

  private clear(args: Record<string, unknown>) {
    const id = typeof args.id === 'string' ? args.id : undefined;
    const invariant = typeof args.invariant === 'string' ? args.invariant : undefined;
    const target = this.oathLedger.find(
      (record) =>
        record.status !== 'cleared' &&
        !this.isExpired(record) &&
        ((id && record.id === id) ||
          (invariant &&
            record.invariant.toLowerCase().includes(invariant.toLowerCase()))),
    );
    if (!target) {
      return {
        action: 'oath.clear.noop',
        cleared: false,
        activeOathCount: this.activeOathCount(),
        nonAuthoritative: true,
      };
    }
    if (target.bindingStrength === 'mast-bound') {
      return {
        action: 'oath.clear.blocked',
        cleared: false,
        oath: target,
        activeOathCount: this.activeOathCount(),
        theoremRef: OATH_THEOREM_REF,
        nonAuthoritative: true,
        bindingStrength: target.bindingStrength,
        reason:
          'mast-bound oath cannot be cleared by the tempted runtime; wait for expiry or use the external user confirmation path',
      };
    }
    target.status = 'cleared';
    target.clearedAt = this.now();
    target.clearReason =
      typeof args.reason === 'string' ? args.reason : 'operator-requested-clear';
    return {
      action: 'oath.cleared',
      cleared: true,
      oath: target,
      telemetryEvent: this.telemetryEvent('oath.cleared', target),
      nonAuthoritative: true,
      auditPreserved: true,
    };
  }

  private audit(args: Record<string, unknown>) {
    const status = typeof args.status === 'string' ? args.status : 'all';
    const oaths = (
      status === 'all'
        ? this.oathLedger
        : this.oathLedger.filter((record) => this.effectiveStatus(record) === status)
    ).map((record) => ({
      ...record,
      effectiveStatus: this.effectiveStatus(record),
      remainingMs: Math.max(0, record.expiresAt - this.now()),
    }));
    return {
      action: 'oath.audit',
      oaths,
      activeOathCount: this.activeOathCount(),
      telemetryEvents: this.auditTelemetryEvents(),
      theoremRef: OATH_THEOREM_REF,
      nonAuthoritative: true,
    };
  }

  private activeOathCount(): number {
    return this.oathLedger.filter(
      (record) => record.status === 'active' && !this.isExpired(record),
    ).length;
  }

  private isExpired(record: ThothOathRecord): boolean {
    return record.status !== 'cleared' && this.now() >= record.expiresAt;
  }

  private effectiveStatus(record: ThothOathRecord): ThothOathRecord['status'] | 'expired' {
    return this.isExpired(record) ? 'expired' : record.status;
  }

  private telemetryEvent(
    kind: ThothOathTelemetryEventKind,
    record?: ThothOathRecord,
  ): ThothOathTelemetryEvent {
    const observedAt = this.now();
    return {
      kind,
      ...(record
        ? {
            oathId: record.id,
            invariant: record.invariant,
            schedule: record.schedule,
            effectiveStatus: this.effectiveStatus(record),
          }
        : {}),
      activeOathCount: this.activeOathCount(),
      theoremRef: OATH_THEOREM_REF,
      nonAuthoritative: true,
      observedAt,
    };
  }

  private auditTelemetryEvents(): ThothOathTelemetryEvent[] {
    const expired = this.oathLedger
      .filter((record) => this.isExpired(record))
      .map((record) => this.telemetryEvent('oath.expired', record));
    return [this.telemetryEvent('oath.audit'), ...expired];
  }

  private async hydrateLedger(): Promise<void> {
    if (this.hydrated || !this.ledgerStore) {
      this.hydrated = true;
      return;
    }
    const records = await this.ledgerStore.load();
    this.oathLedger.splice(0, this.oathLedger.length, ...records.map((record) => ({ ...record })));
    this.hydrated = true;
  }

  private async persistLedger(): Promise<void> {
    if (!this.ledgerStore) return;
    await this.ledgerStore.save(this.oathLedger.map((record) => ({ ...record })));
  }

  private result(
    toolCallId: string,
    toolName: string,
    args: Record<string, unknown>,
    result: unknown,
    startedAt: number,
  ): ToolExecutionResult {
    return {
      toolCallId,
      toolName,
      args,
      result,
      success: true,
      durationMs: this.now() - startedAt,
    };
  }
}

/**
 * Creates the Thoth Oath Self Management Mcp Client.
 */
export function createThothOathSelfManagementMcpClient(
  options: ThothOathToolOptions = {},
): ThothOathSelfManagementMcpClient {
  return new ThothOathSelfManagementMcpClient(options);
}
