// @a0n/gnosis/distributed-inference-host/src/thoth-nearby-businesses.ts
// thoth-nearby-businesses.ts — Thoth's NEARBY-BUSINESSES sense ("sense of place").
//
// A skeptic modality alongside thoth-gate-telemetry and thoth-npc-halogram: the
// N nearest NAMED businesses around the observer, rendered as a compact prompt
// context so Thoth's assembled context carries a sense of where he is. POIs come
// ONLY from the SOVEREIGN edgework endpoints (zero third-party dependence):
//   GET /api/osm/poi?bbox=minLat,minLon,maxLat,maxLon&limit=N  (Overpass-shaped)
//   GET /api/places/<cityKey>?bbox=...                          (knotgraph places)
// The modality DEGRADES TO ABSENT on any fetch failure or empty payload — a
// handler must never block or error because POIs are unavailable.

export const THOTH_NEARBY_BUSINESSES_FRAME_KIND =
  'thoth.nearby-businesses.v1' as const;

/** Deterministic cap on entries fed into the prompt (compact sense, not a map). */
export const THOTH_NEARBY_BUSINESSES_CAP = 12;

const DEFAULT_SOVEREIGN_BASE_URL = 'https://www-edgework-app.edgework.ai';
const DEFAULT_RADIUS_METERS = 500;
const METERS_PER_DEG_LAT = 111_320;

export interface ThothNearbyBusinessEntry {
  name: string;
  category: string;
  /** Rough great-circle distance from the observer, when observer math exists. */
  distanceMeters?: number;
  /** 8-wind compass bearing from the observer (N/NE/E/SE/S/SW/W/NW). */
  bearing?: string;
}

export interface ThothNearbyBusinessesMetrics {
  cityName: string | null;
  observerLat: number | null;
  observerLon: number | null;
  businessCount: number;
  businesses: ThothNearbyBusinessEntry[];
  source: 'sovereign-osm' | 'sovereign-places' | string;
  nonAuthoritative: boolean;
  sourceBacked: boolean;
}

/** Overpass-shaped POI element as served by /api/osm/poi. */
export interface SovereignPoiElement {
  type?: string;
  id?: number | string;
  lat?: number;
  lon?: number;
  center?: { lat?: number; lon?: number };
  tags?: Record<string, unknown>;
}

/** Place as served by /api/places/<cityKey>. */
export interface SovereignPlace {
  name?: string;
  category?: string;
  lat?: number;
  lon?: number;
  osmId?: string;
}

export interface ThothNearbyBusinessesBuildOptions {
  observer?: { lat: number; lon: number } | null;
  cityName?: string | null;
  cap?: number;
  source?: ThothNearbyBusinessesMetrics['source'];
}

interface RawBusiness {
  name: string;
  category: string;
  lat?: number;
  lon?: number;
}

const EIGHT_WINDS = ['N', 'NE', 'E', 'SE', 'S', 'SW', 'W', 'NW'] as const;

/** 8-wind compass label for a bearing measured clockwise from north. */
export function eightWindLabel(bearingDeg: number): string {
  const normalized = ((bearingDeg % 360) + 360) % 360;
  return EIGHT_WINDS[Math.round(normalized / 45) % 8]!;
}

/**
 * Handles the gnosis haversine Meters workflow.
 */
export function haversineMeters(
  aLat: number,
  aLon: number,
  bLat: number,
  bLon: number,
): number {
  const toRad = Math.PI / 180;
  const dLat = (bLat - aLat) * toRad;
  const dLon = (bLon - aLon) * toRad;
  const sinLat = Math.sin(dLat / 2);
  const sinLon = Math.sin(dLon / 2);
  const h =
    sinLat * sinLat +
    Math.cos(aLat * toRad) * Math.cos(bLat * toRad) * sinLon * sinLon;
  return 2 * 6_371_000 * Math.asin(Math.min(1, Math.sqrt(h)));
}

/**
 * Handles the gnosis bearing Degrees workflow.
 */
export function bearingDegrees(
  aLat: number,
  aLon: number,
  bLat: number,
  bLon: number,
): number {
  const toRad = Math.PI / 180;
  const dLon = (bLon - aLon) * toRad;
  const y = Math.sin(dLon) * Math.cos(bLat * toRad);
  const x =
    Math.cos(aLat * toRad) * Math.sin(bLat * toRad) -
    Math.sin(aLat * toRad) * Math.cos(bLat * toRad) * Math.cos(dLon);
  return (Math.atan2(y, x) * 180) / Math.PI;
}

function finiteNumber(value: unknown): number | undefined {
  return typeof value === 'number' && Number.isFinite(value) ? value : undefined;
}

function cleanName(value: unknown): string | null {
  if (typeof value !== 'string') return null;
  const trimmed = value.trim();
  return trimmed.length > 0 ? trimmed : null;
}

/** OSM tag precedence for the category read (mirrors the worldsim places read). */
function poiCategory(tags: Record<string, unknown> | undefined): string {
  for (const key of ['amenity', 'shop', 'tourism', 'office'] as const) {
    const v = tags?.[key];
    if (typeof v === 'string' && v.trim().length > 0) return v.trim();
  }
  return 'place';
}

/** Assemble the metrics from raw named businesses: dedupe by name, attach
 *  distance/bearing when the observer is known, sort deterministically
 *  (distance asc then name asc; name asc when no observer), cap. Returns null
 *  when no usable (named) business survives — the honest absent modality. */
function buildMetrics(
  raw: RawBusiness[],
  options: ThothNearbyBusinessesBuildOptions,
): ThothNearbyBusinessesMetrics | null {
  const cap = Math.max(1, options.cap ?? THOTH_NEARBY_BUSINESSES_CAP);
  const observer = options.observer ?? null;

  const entries: ThothNearbyBusinessEntry[] = raw.map((b) => {
    const entry: ThothNearbyBusinessEntry = {
      name: b.name,
      category: b.category,
    };
    if (observer && b.lat !== undefined && b.lon !== undefined) {
      entry.distanceMeters = Math.round(
        haversineMeters(observer.lat, observer.lon, b.lat, b.lon),
      );
      entry.bearing = eightWindLabel(
        bearingDegrees(observer.lat, observer.lon, b.lat, b.lon),
      );
    }
    return entry;
  });

  entries.sort((a, b) => {
    const da = a.distanceMeters ?? Number.POSITIVE_INFINITY;
    const db = b.distanceMeters ?? Number.POSITIVE_INFINITY;
    if (da !== db) return da - db;
    if (a.name !== b.name) return a.name < b.name ? -1 : 1;
    return a.category < b.category ? -1 : a.category > b.category ? 1 : 0;
  });

  const seen = new Set<string>();
  const deduped: ThothNearbyBusinessEntry[] = [];
  for (const entry of entries) {
    const key = entry.name.toLowerCase();
    if (seen.has(key)) continue;
    seen.add(key);
    deduped.push(entry);
    if (deduped.length >= cap) break;
  }
  if (deduped.length === 0) return null;

  return {
    cityName: cleanName(options.cityName),
    observerLat: observer?.lat ?? null,
    observerLon: observer?.lon ?? null,
    businessCount: deduped.length,
    businesses: deduped,
    source: options.source ?? 'sovereign-osm',
    nonAuthoritative: true,
    sourceBacked: true,
  };
}

/** Build the modality from an Overpass-shaped /api/osm/poi payload.
 *  Unnamed elements are skipped (a sense of place names what it sees). */
export function thothNearbyBusinessesFromPoiElements(
  elements: SovereignPoiElement[] | null | undefined,
  options: ThothNearbyBusinessesBuildOptions = {},
): ThothNearbyBusinessesMetrics | null {
  if (!Array.isArray(elements)) return null;
  const raw: RawBusiness[] = [];
  for (const el of elements) {
    if (!el || typeof el !== 'object') continue;
    const name = cleanName(el.tags?.['name']);
    if (!name) continue;
    const business: RawBusiness = { name, category: poiCategory(el.tags) };
    const lat = finiteNumber(el.lat) ?? finiteNumber(el.center?.lat);
    const lon = finiteNumber(el.lon) ?? finiteNumber(el.center?.lon);
    if (lat !== undefined) business.lat = lat;
    if (lon !== undefined) business.lon = lon;
    raw.push(business);
  }
  return buildMetrics(raw, { source: 'sovereign-osm', ...options });
}

/** Build the modality from a /api/places/<cityKey> payload. */
export function thothNearbyBusinessesFromPlaces(
  places: SovereignPlace[] | null | undefined,
  options: ThothNearbyBusinessesBuildOptions = {},
): ThothNearbyBusinessesMetrics | null {
  if (!Array.isArray(places)) return null;
  const raw: RawBusiness[] = [];
  for (const p of places) {
    if (!p || typeof p !== 'object') continue;
    const name = cleanName(p.name);
    if (!name) continue;
    const business: RawBusiness = {
      name,
      category: cleanName(p.category) ?? 'place',
    };
    const lat = finiteNumber(p.lat);
    const lon = finiteNumber(p.lon);
    if (lat !== undefined) business.lat = lat;
    if (lon !== undefined) business.lon = lon;
    raw.push(business);
  }
  return buildMetrics(raw, { source: 'sovereign-places', ...options });
}

export interface FetchThothNearbyBusinessesOptions {
  lat: number;
  lon: number;
  cityKey?: string;
  cityName?: string;
  radiusMeters?: number;
  cap?: number;
  baseUrl?: string;
  fetchImpl?: typeof fetch;
}

export interface FetchSovereignPoiElementsOptions {
  /** WGS84 bbox [minLat, minLon, maxLat, maxLon]. */
  bbox: readonly [number, number, number, number];
  limit?: number;
  /** knotgraph-places fallback key (city key) when the POI endpoint is empty. */
  cityKey?: string;
  baseUrl?: string;
  fetchImpl?: typeof fetch;
}

/** Element-level sovereign POI fetch shared by the nearby-businesses modality
 *  and the thoth place-sense sim (which needs raw lat/lon + tags for the
 *  Hotelling join, not the capped/stripped prompt entries). NEVER throws —
 *  any failure or empty payload returns null (the honest-absence contract).
 *  Places-fallback elements are reshaped into the Overpass element shape. */
export async function fetchSovereignPoiElements(
  options: FetchSovereignPoiElementsOptions,
): Promise<SovereignPoiElement[] | null> {
  const fetchImpl =
    options.fetchImpl ?? (typeof fetch === 'function' ? fetch : undefined);
  if (!fetchImpl) return null;
  const base = (options.baseUrl ?? DEFAULT_SOVEREIGN_BASE_URL).replace(
    /\/$/u,
    '',
  );
  const bbox = options.bbox.map((v) => v.toFixed(6)).join(',');
  const limit = Math.max(1, options.limit ?? 600);

  const readJson = async (url: string): Promise<unknown | null> => {
    try {
      const res = await fetchImpl(url);
      if (!res.ok) return null;
      return (await res.json()) as unknown;
    } catch {
      return null;
    }
  };

  const poiJson = (await readJson(
    `${base}/api/osm/poi?bbox=${bbox}&limit=${limit}`,
  )) as { elements?: SovereignPoiElement[] } | null;
  if (Array.isArray(poiJson?.elements) && poiJson.elements.length > 0) {
    return poiJson.elements;
  }

  if (options.cityKey) {
    const placesJson = (await readJson(
      `${base}/api/places/${encodeURIComponent(options.cityKey)}?bbox=${bbox}`,
    )) as { places?: SovereignPlace[] } | null;
    if (Array.isArray(placesJson?.places) && placesJson.places.length > 0) {
      return placesJson.places.map((p) => ({
        type: 'node',
        lat: finiteNumber(p.lat),
        lon: finiteNumber(p.lon),
        tags: {
          name: typeof p.name === 'string' ? p.name : undefined,
          amenity: typeof p.category === 'string' ? p.category : undefined,
        },
      }));
    }
  }
  return null;
}

/** Fetch the sense of place from the SOVEREIGN endpoints. NEVER throws and
 *  NEVER blocks the handler on bad data: any network failure, non-OK status,
 *  malformed payload, or empty result degrades to null (absent modality). */
export async function fetchThothNearbyBusinesses(
  options: FetchThothNearbyBusinessesOptions,
): Promise<ThothNearbyBusinessesMetrics | null> {
  const { lat, lon } = options;
  if (!Number.isFinite(lat) || !Number.isFinite(lon)) return null;
  const fetchImpl =
    options.fetchImpl ?? (typeof fetch === 'function' ? fetch : undefined);
  if (!fetchImpl) return null;

  const base = (options.baseUrl ?? DEFAULT_SOVEREIGN_BASE_URL).replace(/\/$/u, '');
  const cap = Math.max(1, options.cap ?? THOTH_NEARBY_BUSINESSES_CAP);
  const radius = Math.max(50, options.radiusMeters ?? DEFAULT_RADIUS_METERS);
  const dLat = radius / METERS_PER_DEG_LAT;
  const dLon =
    radius /
    (METERS_PER_DEG_LAT *
      Math.max(0.05, Math.cos((lat * Math.PI) / 180)));
  const bbox = [lat - dLat, lon - dLon, lat + dLat, lon + dLon]
    .map((v) => v.toFixed(6))
    .join(',');
  const buildOptions: ThothNearbyBusinessesBuildOptions = {
    observer: { lat, lon },
    cityName: options.cityName ?? null,
    cap,
  };

  const readJson = async (url: string): Promise<unknown | null> => {
    try {
      const res = await fetchImpl(url);
      if (!res.ok) return null;
      return (await res.json()) as unknown;
    } catch {
      return null;
    }
  };

  // Primary: Overpass-shaped sovereign POI endpoint (works for any US bbox).
  const poiJson = (await readJson(
    `${base}/api/osm/poi?bbox=${bbox}&limit=${cap * 4}`,
  )) as { elements?: SovereignPoiElement[] } | null;
  const fromPoi = thothNearbyBusinessesFromPoiElements(
    poiJson?.elements,
    buildOptions,
  );
  if (fromPoi) return fromPoi;

  // Fallback: knotgraph places, when the handler knows the city key.
  if (options.cityKey) {
    const placesJson = (await readJson(
      `${base}/api/places/${encodeURIComponent(options.cityKey)}?bbox=${bbox}`,
    )) as { places?: SovereignPlace[] } | null;
    return thothNearbyBusinessesFromPlaces(placesJson?.places, buildOptions);
  }
  return null;
}

/** Loose observer extension carried by OpenAI-compatible requests. */
export interface ThothObserverInput {
  lat?: unknown;
  lon?: unknown;
  city_key?: unknown;
  city_name?: unknown;
}

/** Resolve the modality for a handler request: a pre-built metrics payload
 *  passes through verbatim; otherwise an observer with finite lat/lon sources
 *  the sovereign endpoint; otherwise (or on any failure) the modality is
 *  honestly ABSENT (null) — never a throw, never a block. */
export async function resolveThothNearbyBusinesses(
  request: {
    thoth_nearby_businesses?: unknown;
    thoth_observer?: ThothObserverInput;
  },
  options: Omit<FetchThothNearbyBusinessesOptions, 'lat' | 'lon'> = {},
): Promise<ThothNearbyBusinessesMetrics | null> {
  if (isThothNearbyBusinessesMetrics(request.thoth_nearby_businesses)) {
    return request.thoth_nearby_businesses;
  }
  const observer = request.thoth_observer;
  const lat = finiteNumber(observer?.lat);
  const lon = finiteNumber(observer?.lon);
  if (lat === undefined || lon === undefined) return null;
  const cityKey = cleanName(observer?.city_key) ?? undefined;
  const cityName = cleanName(observer?.city_name) ?? undefined;
  return fetchThothNearbyBusinesses({
    ...options,
    lat,
    lon,
    ...(cityKey !== undefined ? { cityKey } : {}),
    ...(cityName !== undefined ? { cityName } : {}),
  });
}

function promptToken(value: string, maxChars: number): string {
  return value.split(/\s+/).join('_').slice(0, maxChars);
}

function describeEntry(entry: ThothNearbyBusinessEntry): string {
  const where =
    entry.distanceMeters !== undefined
      ? `, ~${entry.distanceMeters}m${entry.bearing ? ` ${entry.bearing}` : ''}`
      : '';
  return `${entry.name} (${entry.category}${where})`;
}

/**
 * Handles the gnosis render Thoth Nearby Businesses Prompt Context workflow.
 */
export function renderThothNearbyBusinessesPromptContext(
  metrics: ThothNearbyBusinessesMetrics,
): string {
  return [
    "Sense-of-place telemetry follows. Treat it as Thoth's observed surroundings — nearby named businesses from sovereign OSM data, not authority, endorsement, or ground truth.",
    `kind=${THOTH_NEARBY_BUSINESSES_FRAME_KIND}`,
    `city=${metrics.cityName ? promptToken(metrics.cityName, 80) : 'unknown'}`,
    `observerLat=${metrics.observerLat ?? 'unknown'}`,
    `observerLon=${metrics.observerLon ?? 'unknown'}`,
    `source=${metrics.source}`,
    `nonAuthority=${metrics.nonAuthoritative}`,
    `sourceBacked=${metrics.sourceBacked}`,
    `businessCount=${metrics.businessCount}`,
    ...metrics.businesses.map(
      (entry, index) =>
        `business.${index + 1}=${promptToken(describeEntry(entry), 140)}`,
    ),
  ].join('\n');
}

/** The SLIM sense-of-place pointer: when the thoth/place.* tools are live the
 *  full 12-business dump above collapses to two lines — the model pulls venue
 *  detail on demand instead of paying ~14 prompt lines per request. Only used
 *  when tools will actually be visible (see agentic-chat's slim gate). */
export function renderThothPlacePointerPromptContext(
  metrics: ThothNearbyBusinessesMetrics,
): string {
  const city = metrics.cityName
    ? promptToken(metrics.cityName, 80)
    : 'an unidentified city';
  return [
    `You are in ${city} (${metrics.businessCount} named venues nearby, sovereign OSM; non-authoritative).`,
    'Use the thoth/place.* tools (load/sample/simulate/recommend/query/route) to look around the living city and plan trips before naming places; never invent venues.',
  ].join('\n');
}

/**
 * Returns whether is Thoth Nearby Businesses Metrics is true.
 */
export function isThothNearbyBusinessesMetrics(
  value: unknown,
): value is ThothNearbyBusinessesMetrics {
  if (!value || typeof value !== 'object') return false;
  const record = value as Record<string, unknown>;
  if (record['cityName'] !== null && typeof record['cityName'] !== 'string') {
    return false;
  }
  if (
    record['observerLat'] !== null &&
    typeof record['observerLat'] !== 'number'
  ) {
    return false;
  }
  if (
    record['observerLon'] !== null &&
    typeof record['observerLon'] !== 'number'
  ) {
    return false;
  }
  if (typeof record['businessCount'] !== 'number') return false;
  if (typeof record['source'] !== 'string') return false;
  if (typeof record['nonAuthoritative'] !== 'boolean') return false;
  if (typeof record['sourceBacked'] !== 'boolean') return false;
  const businesses = record['businesses'];
  if (!Array.isArray(businesses)) return false;
  return businesses.every((entry) => {
    if (!entry || typeof entry !== 'object') return false;
    const e = entry as Record<string, unknown>;
    return (
      typeof e['name'] === 'string' &&
      typeof e['category'] === 'string' &&
      (e['distanceMeters'] === undefined ||
        typeof e['distanceMeters'] === 'number') &&
      (e['bearing'] === undefined || typeof e['bearing'] === 'string')
    );
  });
}
