// @a0n/gnosis/distributed-inference-host/src/thoth-gate-telemetry.ts
export const AEON_STREAM_PNEUMA_INTERNAL = 2002 as const;
export const THOTH_GATE_TELEMETRY_FRAME_KIND =
  'thoth.gate-telemetry.v1' as const;

export const THOTH_GATE_KINDS = [
  'horoscope',
  'oath',
  'prana',
  'ki',
  'chi',
  'tai-chi',
  'mudra',
  'cognitive',
  'aura',
  'safety',
  'audio-shield',
] as const;

export type ThothGateKind = (typeof THOTH_GATE_KINDS)[number];

export interface ThothGateScore {
  gate: ThothGateKind;
  aligned: number;
  blocked: number;
  overconcentrated: number;
  underfed: number;
  readyToHandoff: number;
  perspective: string;
  theoremRef: string;
  provenance: readonly string[];
  failureResidue: string;
}

export interface ThothGateTelemetryMetrics {
  scores: readonly ThothGateScore[];
  observationWithoutIntervention: boolean;
  nonAuthoritative: boolean;
  sourceBacked: boolean;
  gateBoundariesPreserved: boolean;
}

export interface ThothGateTelemetryFrameWire {
  mediaType: {
    Abstract: typeof THOTH_GATE_TELEMETRY_FRAME_KIND;
  };
  level: number;
  phase: number;
  deficit: number;
  barcode: {
    beta1: number;
    lifespan: number;
    dilation: number;
    density: number;
  };
  data: ThothGateTelemetryMetrics;
  attributes: Record<string, string>;
  sourceId: string;
}

const perspectiveByGate: Record<ThothGateKind, string> = {
  horoscope: 'systemic-alignment',
  oath: 'stability-binding',
  prana: 'body-gate-damping',
  ki: 'preincident-defense',
  chi: 'environmental-flow',
  'tai-chi': 'moving-polarity-balance',
  mudra: 'gesture-body-gate',
  cognitive: 'system-two-handoff',
  aura: 'relational-weather',
  safety: 'policy-envelope',
  'audio-shield': 'incoming-manipulation-gate',
};

function score(
  gate: ThothGateKind,
  aligned: number,
  blocked: number,
  overconcentrated: number,
  underfed: number,
  readyToHandoff: number,
  theoremRef: string,
  provenance: readonly string[],
  failureResidue: string,
): ThothGateScore {
  return {
    gate,
    aligned,
    blocked,
    overconcentrated,
    underfed,
    readyToHandoff,
    perspective: perspectiveByGate[gate],
    theoremRef,
    provenance,
    failureResidue,
  };
}

/**
 * Handles the gnosis canonical Thoth Gate Telemetry Metrics workflow.
 */
export function canonicalThothGateTelemetryMetrics(): ThothGateTelemetryMetrics {
  return {
    scores: [
      score(
        'horoscope',
        0.7,
        0.1,
        0.2,
        0,
        0.4,
        'Gnosis.Witnesses.Folklore.ThothZodiacTelemetryGateWitness.thoth_zodiac_telemetry_gate_witness',
        [
          'Gnosis.Witnesses.Folklore.ZodiacAlignmentScoringWitness',
          'Gnosis.Witnesses.Folklore.ZodiacCosmicAlignmentMetricWitness',
        ],
        'prediction/personality authority rejected',
      ),
      score(
        'oath',
        0.85,
        0.05,
        0,
        0.1,
        0.45,
        'Gnosis.Witnesses.Interfaith.OathStabilityGateWitness.oath_stability_gate_witness',
        [
          'Gnosis.Witnesses.Interfaith.MalabarArchdeaconAssemblyResistanceWitness',
          'Gnosis.Witnesses.Folklore.HospitalityCovenantBoundaryWitness',
        ],
        'stability binds operator intent without collapsing agent freedom',
      ),
      score(
        'prana',
        0.8,
        0,
        0,
        0.2,
        0.3,
        'Gnosis.Witnesses.Hindu.ThothHoroscopePranaOptimizationWitness.thoth_horoscope_prana_optimization_witness',
        [
          'Gnosis.Witnesses.Hindu.PranaThothTelemetryDampingWitness',
          'Gnosis.Witnesses.Hindu.PranaFailureStandingWaveWitness',
        ],
        'damping lowers amplitude without erasing viable signal',
      ),
      score(
        'ki',
        0.5,
        0.3,
        0,
        0.1,
        0.6,
        'Gnosis.Witnesses.Folklore.KiObservationalDefenseGateWitness.ki_observational_defense_gate_witness',
        [
          'shared-ui/src/services/metacognition/kiPremetacog.ts',
          'Gnosis.Bridges.KiFrfTrainingQueueKernelBridge',
        ],
        'action requires timing order and chance-gate evidence',
      ),
      score(
        'chi',
        0.6,
        0.2,
        0,
        0.2,
        0.2,
        'Gnosis.FengShui.five_elements_pleromatic_closure',
        ['Gnosis.FengShuiTopology'],
        'environmental flow is not semantic authority',
      ),
      score(
        'tai-chi',
        0.75,
        0.05,
        0,
        0.2,
        0.35,
        'Gnosis.Witnesses.Folklore.TaiChiYinYangBalanceGateWitness.tai_chi_yin_yang_balance_gate_witness',
        [
          'Gnosis.Witnesses.Tao.TaoTeChingLowPlaceWuWeiWitness',
          'Gnosis.Witnesses.Hermetic.KybalionRhythmNeutralizationWitness',
        ],
        'moving balance is tracked without force escalation',
      ),
      score(
        'mudra',
        0.7,
        0,
        0,
        0.3,
        0.25,
        'Gnosis.Witnesses.Hindu.YogicMudraBreathGateWitness.yogic_gate_telemetry_bridge_sound',
        [
          'Gnosis.MudraTopology',
          'docs/ebooks/source-texts/bhagavad-gita-arnold.txt',
        ],
        'gesture remains body-gate evidence, not doctrine',
      ),
      score(
        'cognitive',
        0.45,
        0.35,
        0.05,
        0.15,
        0.55,
        'shared-ui.cognitive-gate',
        ['shared-ui/src/hooks/useCognitiveGate.ts'],
        'system-two handoff remains advisory',
      ),
      score(
        'aura',
        0.5,
        0.25,
        0,
        0.25,
        0.2,
        'shared-ui.aura-detect',
        ['shared-ui/src/edge/tools/aura-detect/AuraDetectPageContent.tsx'],
        'relational weather is observation-only',
      ),
      score(
        'safety',
        0.65,
        0.3,
        0,
        0.05,
        0.5,
        'shared-ui.senju-safety-gate',
        ['shared-ui/src/services/senju/policy/safetyGate.ts'],
        'policy envelope can force pause without claiming truth',
      ),
      score(
        'audio-shield',
        0.4,
        0.45,
        0,
        0.15,
        0.5,
        'shared-ui.streaming-audio-gate',
        ['shared-ui/src/services/shield/streamingAudioGate.ts'],
        'incoming manipulation gate is defensive only',
      ),
    ],
    observationWithoutIntervention: true,
    nonAuthoritative: true,
    sourceBacked: true,
    gateBoundariesPreserved: true,
  };
}

/**
 * Handles the gnosis thoth Gate Telemetry Total Pressure workflow.
 */
export function thothGateTelemetryTotalPressure(
  metrics: ThothGateTelemetryMetrics,
): number {
  return metrics.scores.reduce(
    (sum, gate) => sum + gate.blocked + gate.overconcentrated + gate.underfed,
    0,
  );
}

/**
 * Handles the gnosis thoth Gate Telemetry Ready Handoff Count workflow.
 */
export function thothGateTelemetryReadyHandoffCount(
  metrics: ThothGateTelemetryMetrics,
): number {
  return metrics.scores.filter((gate) => gate.readyToHandoff > 0).length;
}

/**
 * Handles the gnosis encode Thoth Gate Telemetry Frame workflow.
 */
export function encodeThothGateTelemetryFrame(
  sourceId: string,
  metrics: ThothGateTelemetryMetrics = canonicalThothGateTelemetryMetrics(),
): ThothGateTelemetryFrameWire {
  return {
    mediaType: { Abstract: THOTH_GATE_TELEMETRY_FRAME_KIND },
    level: 10,
    phase: 2,
    deficit: thothGateTelemetryTotalPressure(metrics),
    barcode: {
      beta1: Math.max(1, metrics.scores.length),
      lifespan: 1,
      dilation: 1,
      density: 1,
    },
    data: metrics,
    attributes: {
      aeon_stream_id: String(AEON_STREAM_PNEUMA_INTERNAL),
      frame_role: 'consciousness',
      thoth_frame_kind: THOTH_GATE_TELEMETRY_FRAME_KIND,
      gate_score_count: String(metrics.scores.length),
      gate_names: metrics.scores.map((gate) => gate.gate).join('|'),
      gate_total_pressure: String(thothGateTelemetryTotalPressure(metrics)),
      gate_ready_handoff_count: String(
        thothGateTelemetryReadyHandoffCount(metrics),
      ),
      observation_without_intervention: String(
        metrics.observationWithoutIntervention,
      ),
      non_authoritative: String(metrics.nonAuthoritative),
      source_backed: String(metrics.sourceBacked),
      gate_boundaries_preserved: String(metrics.gateBoundariesPreserved),
      synthetic_gnosis_admissible: 'true',
      no_source_authority_claim: 'true',
    },
    sourceId,
  };
}

/**
 * Returns whether is Thoth Gate Telemetry Frame is true.
 */
export function isThothGateTelemetryFrame(
  frame: Pick<ThothGateTelemetryFrameWire, 'mediaType' | 'attributes'>,
): boolean {
  return (
    frame.mediaType.Abstract === THOTH_GATE_TELEMETRY_FRAME_KIND &&
    frame.attributes.aeon_stream_id === String(AEON_STREAM_PNEUMA_INTERNAL)
  );
}

/**
 * Handles the gnosis render Thoth Gate Telemetry Prompt Context workflow.
 */
export function renderThothGateTelemetryPromptContext(
  metrics: ThothGateTelemetryMetrics,
): string {
  const gateRows = metrics.scores
    .map(
      (gate) =>
        `- ${gate.gate}: aligned=${gate.aligned.toFixed(2)} pressure=${(
          gate.blocked +
          gate.overconcentrated +
          gate.underfed
        ).toFixed(2)} ready=${gate.readyToHandoff.toFixed(2)} perspective=${
          gate.perspective
        } residue=${gate.failureResidue}`,
    )
    .join('\n');

  return `Thoth gate telemetry follows. Treat it as bounded observation, not authority or diagnosis.
Prompt/nudge/action contract: prompts may cite gate state; nudges may use blocked/underfed gates as next-step hints; actions must remain non-authoritative and source-backed.
Flags: observationOnly=${metrics.observationWithoutIntervention} nonAuthority=${metrics.nonAuthoritative} sourceBacked=${metrics.sourceBacked} boundariesPreserved=${metrics.gateBoundariesPreserved}
${gateRows}`;
}

/**
 * Handles the gnosis thoth Gate Telemetry Action Hint workflow.
 */
export function thothGateTelemetryActionHint(
  metrics: ThothGateTelemetryMetrics,
): string {
  const pressureGate = [...metrics.scores].sort(
    (left, right) =>
      right.blocked +
      right.overconcentrated +
      right.underfed -
      (left.blocked + left.overconcentrated + left.underfed),
  )[0];
  const handoffGates = metrics.scores
    .filter((gate) => gate.readyToHandoff > 0)
    .map((gate) => gate.gate)
    .join('|');
  return `highestPressure=${pressureGate?.gate ?? 'none'} readyHandoffs=${
    handoffGates || 'none'
  } totalPressure=${thothGateTelemetryTotalPressure(metrics).toFixed(2)}`;
}
