// @a0n/gnosis/distributed-inference-host/src/thoth-npc-halogram.ts
export const THOTH_NPC_HALOGRAM_FRAME_KIND = 'thoth.npc-halogram.v1' as const;

export interface ThothNpcHalogramMetrics {
  npcId: string;
  displayName: string;
  projectionSource: 'halogram' | 'hologram' | string;
  bestSelfDescription: string;
  voiceTone: 'warm' | 'direct' | 'gentle' | 'encouraging' | 'balanced' | string;
  voiceFormality: 'formal' | 'casual' | 'adaptive' | string;
  voiceDirectness: number;
  voiceEmotionality: number;
  bodyPresentation: 'man' | 'woman' | 'androgynous' | 'adaptive' | string;
  bodyBuild: string;
  bodyGestureProfile: string;
  bodyPresenceIntensity: number;
  calibratedConfidenceCount: number;
  overallCoherence: number;
  remainingTensionCount: number;
  biasDetectedCount: number;
  correctionAppliedCount: number;
  growthAreaCount: number;
  blindSpotCount: number;
  dataQualityScore: number;
  confidenceInProjection: number;
  aggregateLayersProcessed: number;
  aggregateCorrectionsApplied: number;
  observationWithoutIntervention: boolean;
  nonAuthoritative: boolean;
  sourceBacked: boolean;
  optimalSelfProjection: boolean;
}

export interface SharedUiHalogramSummaryInput {
  userId?: string;
  npcId?: string;
  displayName?: string;
  calibratedConfidences?: Map<unknown, unknown> | Record<string, unknown> | unknown[];
  coherenceAssessment?: {
    overallCoherence?: number;
    remainingTensions?: unknown[];
  };
  biasCorrections?: {
    biasesDetected?: unknown[];
    correctionsApplied?: unknown[];
  };
  optimizedProfile?: {
    growthAreas?: unknown[];
    blindSpots?: unknown[];
    bestSelfDescription?: string;
  };
  voice?: {
    tone?: string;
    formality?: string;
    directness?: number;
    emotionality?: number;
  };
  body?: {
    presentation?: string;
    build?: string;
    gestureProfile?: string;
    presenceIntensity?: number;
  };
  meta?: {
    dataQualityScore?: number;
    confidenceInProjection?: number;
  };
  aggregateMetacognition?: {
    meta?: {
      layersProcessed?: number;
      totalCorrectionsApplied?: number;
    };
  };
}

/**
 * Handles the gnosis thoth Npc Halogram Metrics From Shared Ui Halogram workflow.
 */
export function thothNpcHalogramMetricsFromSharedUiHalogram(
  input: SharedUiHalogramSummaryInput,
): ThothNpcHalogramMetrics {
  const calibratedConfidenceCount = collectionSize(input.calibratedConfidences);
  const aggregateLayersProcessed =
    finiteNumber(input.aggregateMetacognition?.meta?.layersProcessed) ??
    calibratedConfidenceCount;
  const correctionAppliedCount = collectionSize(
    input.biasCorrections?.correctionsApplied,
  );

  return {
    npcId: input.npcId ?? input.userId ?? 'thoth',
    displayName: input.displayName ?? 'Thoth',
    projectionSource: 'halogram',
    bestSelfDescription:
      input.optimizedProfile?.bestSelfDescription ??
      'Bounded optimal-self projection from shared-ui Halogram.',
    voiceTone: input.voice?.tone ?? 'balanced',
    voiceFormality: input.voice?.formality ?? 'adaptive',
    voiceDirectness: clamp01(finiteNumber(input.voice?.directness) ?? 0.5),
    voiceEmotionality: clamp01(finiteNumber(input.voice?.emotionality) ?? 0.5),
    bodyPresentation: input.body?.presentation ?? 'adaptive',
    bodyBuild: input.body?.build ?? 'adaptive',
    bodyGestureProfile: input.body?.gestureProfile ?? 'measured-scribe',
    bodyPresenceIntensity: clamp01(
      finiteNumber(input.body?.presenceIntensity) ?? 0.5,
    ),
    calibratedConfidenceCount,
    overallCoherence: clamp01(
      finiteNumber(input.coherenceAssessment?.overallCoherence) ?? 0,
    ),
    remainingTensionCount: collectionSize(
      input.coherenceAssessment?.remainingTensions,
    ),
    biasDetectedCount: collectionSize(input.biasCorrections?.biasesDetected),
    correctionAppliedCount,
    growthAreaCount: collectionSize(input.optimizedProfile?.growthAreas),
    blindSpotCount: collectionSize(input.optimizedProfile?.blindSpots),
    dataQualityScore: clamp01(finiteNumber(input.meta?.dataQualityScore) ?? 0),
    confidenceInProjection: clamp01(
      finiteNumber(input.meta?.confidenceInProjection) ?? 0,
    ),
    aggregateLayersProcessed,
    aggregateCorrectionsApplied:
      finiteNumber(input.aggregateMetacognition?.meta?.totalCorrectionsApplied) ??
      correctionAppliedCount,
    observationWithoutIntervention: true,
    nonAuthoritative: true,
    sourceBacked: true,
    optimalSelfProjection: true,
  };
}

/**
 * Handles the gnosis render Thoth Npc Halogram Prompt Context workflow.
 */
export function renderThothNpcHalogramPromptContext(
  metrics: ThothNpcHalogramMetrics,
): string {
  return [
    'Shared-ui Halogram NPC telemetry follows. Treat it as bounded optimal-self projection metadata, not authority, diagnosis, or ground truth.',
    `kind=${THOTH_NPC_HALOGRAM_FRAME_KIND}`,
    `npc=${metrics.npcId}`,
    `display=${metrics.displayName}`,
    `projectionSource=${metrics.projectionSource}`,
    `optimalSelf=${metrics.optimalSelfProjection}`,
    `voiceTone=${metrics.voiceTone}`,
    `voiceFormality=${metrics.voiceFormality}`,
    `voiceDirectness=${metrics.voiceDirectness}`,
    `voiceEmotionality=${metrics.voiceEmotionality}`,
    `bodyPresentation=${metrics.bodyPresentation}`,
    `bodyBuild=${metrics.bodyBuild}`,
    `bodyGesture=${metrics.bodyGestureProfile}`,
    `bodyPresence=${metrics.bodyPresenceIntensity}`,
    `calibratedConfidences=${metrics.calibratedConfidenceCount}`,
    `coherence=${metrics.overallCoherence}`,
    `remainingTensions=${metrics.remainingTensionCount}`,
    `biasesDetected=${metrics.biasDetectedCount}`,
    `correctionsApplied=${metrics.correctionAppliedCount}`,
    `growthAreas=${metrics.growthAreaCount}`,
    `blindSpots=${metrics.blindSpotCount}`,
    `dataQuality=${metrics.dataQualityScore}`,
    `projectionConfidence=${metrics.confidenceInProjection}`,
    `aggregateLayers=${metrics.aggregateLayersProcessed}`,
    `aggregateCorrections=${metrics.aggregateCorrectionsApplied}`,
    `observationOnly=${metrics.observationWithoutIntervention}`,
    `nonAuthority=${metrics.nonAuthoritative}`,
    `sourceBacked=${metrics.sourceBacked}`,
    `bestSelf=${promptToken(metrics.bestSelfDescription, 220)}`,
  ].join('\n');
}

/**
 * Returns whether is Thoth Npc Halogram Metrics is true.
 */
export function isThothNpcHalogramMetrics(
  value: unknown,
): value is ThothNpcHalogramMetrics {
  if (!value || typeof value !== 'object') return false;
  const record = value as Record<string, unknown>;
  return (
    typeof record.npcId === 'string' &&
    typeof record.displayName === 'string' &&
    typeof record.projectionSource === 'string' &&
    typeof record.bestSelfDescription === 'string' &&
    typeof record.voiceTone === 'string' &&
    typeof record.voiceFormality === 'string' &&
    typeof record.bodyPresentation === 'string' &&
    typeof record.bodyBuild === 'string' &&
    typeof record.bodyGestureProfile === 'string' &&
    [
      'voiceDirectness',
      'voiceEmotionality',
      'bodyPresenceIntensity',
      'calibratedConfidenceCount',
      'overallCoherence',
      'remainingTensionCount',
      'biasDetectedCount',
      'correctionAppliedCount',
      'growthAreaCount',
      'blindSpotCount',
      'dataQualityScore',
      'confidenceInProjection',
      'aggregateLayersProcessed',
      'aggregateCorrectionsApplied',
    ].every((key) => typeof record[key] === 'number') &&
    [
      'observationWithoutIntervention',
      'nonAuthoritative',
      'sourceBacked',
      'optimalSelfProjection',
    ].every((key) => typeof record[key] === 'boolean')
  );
}

function collectionSize(value: unknown): number {
  if (Array.isArray(value)) return value.length;
  if (value instanceof Map) return value.size;
  if (value && typeof value === 'object') return Object.keys(value).length;
  return 0;
}

function finiteNumber(value: unknown): number | undefined {
  return typeof value === 'number' && Number.isFinite(value) ? value : undefined;
}

function clamp01(value: number): number {
  return Math.max(0, Math.min(1, value));
}

function promptToken(value: string, maxChars: number): string {
  return value.split(/\s+/).join('_').slice(0, maxChars);
}
