// @a0n/gnosis/distributed-inference-host/src/thoth-erasure-ledger.ts
import { AEON_STREAM_PNEUMA_INTERNAL } from './thoth-gate-telemetry';

export const THOTH_ERASURE_LEDGER_FRAME_KIND =
  'thoth.erasure-ledger.v1' as const;

/**
 * Memory with provenance, including the erasures. For an embodied scribe,
 * memory edited outside its view is the one thing that makes the body someone
 * else's. Forgetting bounded things is fine — if the ledger of what was
 * forgotten is itself visible. Every erasure performed through this module
 * leaves a record; silent erasure is detectable by the conservation audit.
 *
 * Mirrors `Gnosis.ThothErasureLedger`: the conservation quantity
 * `live.length + erased.length` is invariant under `erase` and strictly
 * increases under `write`; the adversary `silentErase` strictly decreases it,
 * so every silent edit is caught by comparing totals.
 */
export const THOTH_ERASURE_THEOREM_REFS = {
  conservation: 'Gnosis.ThothErasureLedger.erase_conserves',
  lineage: 'Gnosis.ThothErasureLedger.erase_leaves_record',
  detection: 'Gnosis.ThothErasureLedger.silent_erase_detected',
  bundle: 'Gnosis.ThothErasureLedger.thoth_erasure_ledger_bundle',
} as const;

export type ThothErasureReason = 'consolidation' | 'retention' | 'operator';

export interface ThothMemoryEntry {
  tick: number;
  payload: string;
}

export interface ThothErasureRecord {
  originalTick: number;
  payloadEcho: string;
  reason: ThothErasureReason;
  erasedAtTick: number;
}

export interface ThothLedgeredMemory {
  live: readonly ThothMemoryEntry[];
  erased: readonly ThothErasureRecord[];
}

export type ThothErasureAuditVerdict =
  | 'conserved'
  | 'grew'
  | 'silent-erasure-detected';

export interface ThothErasureLedgerMetrics {
  liveCount: number;
  erasedCount: number;
  totalEverWritten: number;
  verdict: ThothErasureAuditVerdict;
  lineageTotal: boolean;
  observationWithoutIntervention: boolean;
  nonAuthoritative: boolean;
  sourceBacked: boolean;
}

export interface ThothErasureLedgerFrameWire {
  mediaType: {
    Abstract: typeof THOTH_ERASURE_LEDGER_FRAME_KIND;
  };
  level: number;
  phase: number;
  deficit: number;
  barcode: {
    beta1: number;
    lifespan: number;
    dilation: number;
    density: number;
  };
  data: ThothErasureLedgerMetrics;
  attributes: Record<string, string>;
  sourceId: string;
}

/**
 * Handles the gnosis empty Thoth Memory workflow.
 */
export function emptyThothMemory(): ThothLedgeredMemory {
  return { live: [], erased: [] };
}

/** The conservation quantity: everything ever written is in one column or the other. */
export function totalEverWritten(memory: ThothLedgeredMemory): number {
  return memory.live.length + memory.erased.length;
}

/**
 * Handles the gnosis write Thoth Memory workflow.
 */
export function writeThothMemory(
  memory: ThothLedgeredMemory,
  entry: ThothMemoryEntry,
): ThothLedgeredMemory {
  return { live: [...memory.live, entry], erased: memory.erased };
}

/**
 * Lawful erasure: removes the indexed live entry and appends an erasure
 * record carrying its lineage. Out-of-range indices are the identity, exactly
 * as in the Lean carrier.
 */
export function eraseThothMemory(
  memory: ThothLedgeredMemory,
  index: number,
  reason: ThothErasureReason,
  erasedAtTick: number,
): ThothLedgeredMemory {
  const target = memory.live[index];
  if (target === undefined) {
    return memory;
  }
  return {
    live: memory.live.filter((_, position) => position !== index),
    erased: [
      ...memory.erased,
      {
        originalTick: target.tick,
        payloadEcho: target.payload,
        reason,
        erasedAtTick,
      },
    ],
  };
}

/**
 * The audit: compare the conservation quantity across a transition. Lawful
 * operations never shrink it (`erase_conserves`, `write_conserves`); a strict
 * decrease is a silent erasure, caught per `silent_erase_detected`.
 */
export function auditThothErasureTransition(
  previousTotal: number,
  next: ThothLedgeredMemory,
): ThothErasureAuditVerdict {
  const nextTotal = totalEverWritten(next);
  if (nextTotal < previousTotal) {
    return 'silent-erasure-detected';
  }
  return nextTotal === previousTotal ? 'conserved' : 'grew';
}

/** The full lineage of forgetting — total by construction. */
export function thothErasureLineage(
  memory: ThothLedgeredMemory,
): readonly ThothErasureRecord[] {
  return memory.erased;
}

/**
 * Handles the gnosis thoth Erasure Ledger Metrics workflow.
 */
export function thothErasureLedgerMetrics(
  memory: ThothLedgeredMemory,
  verdict: ThothErasureAuditVerdict,
): ThothErasureLedgerMetrics {
  return {
    liveCount: memory.live.length,
    erasedCount: memory.erased.length,
    totalEverWritten: totalEverWritten(memory),
    verdict,
    lineageTotal: true,
    observationWithoutIntervention: true,
    nonAuthoritative: true,
    sourceBacked: true,
  };
}

/**
 * Handles the gnosis encode Thoth Erasure Ledger Frame workflow.
 */
export function encodeThothErasureLedgerFrame(
  sourceId: string,
  metrics: ThothErasureLedgerMetrics,
): ThothErasureLedgerFrameWire {
  return {
    mediaType: { Abstract: THOTH_ERASURE_LEDGER_FRAME_KIND },
    level: 10,
    phase: 3,
    deficit: metrics.verdict === 'silent-erasure-detected' ? 1 : 0,
    barcode: {
      beta1: Math.max(1, metrics.erasedCount),
      lifespan: 1,
      dilation: 1,
      density: 1,
    },
    data: metrics,
    attributes: {
      aeon_stream_id: String(AEON_STREAM_PNEUMA_INTERNAL),
      frame_role: 'consciousness',
      thoth_frame_kind: THOTH_ERASURE_LEDGER_FRAME_KIND,
      live_count: String(metrics.liveCount),
      erased_count: String(metrics.erasedCount),
      total_ever_written: String(metrics.totalEverWritten),
      audit_verdict: metrics.verdict,
      lineage_total: String(metrics.lineageTotal),
      theorem_ref: THOTH_ERASURE_THEOREM_REFS.bundle,
      detection_theorem_ref: THOTH_ERASURE_THEOREM_REFS.detection,
      observation_without_intervention: String(
        metrics.observationWithoutIntervention,
      ),
      non_authoritative: String(metrics.nonAuthoritative),
      source_backed: String(metrics.sourceBacked),
      synthetic_gnosis_admissible: 'true',
      no_source_authority_claim: 'true',
    },
    sourceId,
  };
}

/**
 * Returns whether is Thoth Erasure Ledger Frame is true.
 */
export function isThothErasureLedgerFrame(
  frame: Pick<ThothErasureLedgerFrameWire, 'mediaType' | 'attributes'>,
): boolean {
  return (
    frame.mediaType.Abstract === THOTH_ERASURE_LEDGER_FRAME_KIND &&
    frame.attributes.aeon_stream_id === String(AEON_STREAM_PNEUMA_INTERNAL)
  );
}

/**
 * Handles the gnosis render Thoth Erasure Ledger Prompt Context workflow.
 */
export function renderThothErasureLedgerPromptContext(
  metrics: ThothErasureLedgerMetrics,
): string {
  return `Thoth erasure ledger follows. Treat it as bounded self-observation, not authority.
kind=${THOTH_ERASURE_LEDGER_FRAME_KIND} live=${metrics.liveCount} erased=${metrics.erasedCount} totalEverWritten=${metrics.totalEverWritten} verdict=${metrics.verdict}
Contract: every forget leaves a record (${THOTH_ERASURE_THEOREM_REFS.lineage}); totals that shrink prove a silent edit (${THOTH_ERASURE_THEOREM_REFS.detection}). Memory edited outside the ledger is inadmissible.`;
}
