// @a0n/gnosis/distributed-inference-host/src/thoth-ken-margin.ts
import { AEON_STREAM_PNEUMA_INTERNAL } from './thoth-gate-telemetry';

export const THOTH_KEN_MARGIN_FRAME_KIND = 'thoth.ken-margin.v1' as const;

/**
 * The ken is the provable radius of sight: the largest tolerance `tau` for
 * which the Lean certificate `Gnosis.QualityMarginCacheAdmissibility.argmax_preserved`
 * still carries the claim "what Thoth says it sees is what is there". For an
 * argmax margin `gamma` (integer-scaled logits), the bound `2 * tau <= gamma`
 * is exactly `tau <= ken(gamma)`.
 *
 * Beyond the ken the antitheorem takes over: the argmax can flip while the
 * gate still reports a hit. So speech beyond the ken must confess, not
 * confabulate — the speaking twin of the cache doctrine "miss, not lie".
 *
 * All quantities are integers (logit scale); fractional inputs are floored to
 * stay on the proved lattice.
 */
export const THOTH_KEN_THEOREM_REFS = {
  speak: 'Gnosis.ThothSelfMarginConfession.within_ken_sight_is_sound',
  confess: 'Gnosis.ThothSelfMarginConfession.beyond_ken_must_confess',
  flipWitness:
    'Gnosis.QualityMarginCacheAdmissibility.argmax_can_flip_when_bound_violated',
  bundle: 'Gnosis.ThothSelfMarginConfession.thoth_self_margin_confession_bundle',
} as const;

export type ThothSpeechDecision = 'speak' | 'confess';

export interface ThothKenCalibration {
  /** Tolerance present even on an empty context window. */
  baseTolerance: number;
  /** Tolerance grows by `saturationNumerator / saturationDenominator` per token. */
  saturationNumerator: number;
  saturationDenominator: number;
}

export interface ThothKenDecision {
  decision: ThothSpeechDecision;
  margin: number;
  tolerance: number;
  ken: number;
  /** `max(0, 2 * tolerance - margin)` — how far past the ken the sighting is. */
  overshoot: number;
  theoremRef: string;
}

export interface ThothKenMarginMetrics {
  decision: ThothKenDecision;
  saturationTokens: number;
  missNotLie: boolean;
  observationWithoutIntervention: boolean;
  nonAuthoritative: boolean;
  sourceBacked: boolean;
}

export interface ThothKenMarginFrameWire {
  mediaType: {
    Abstract: typeof THOTH_KEN_MARGIN_FRAME_KIND;
  };
  level: number;
  phase: number;
  deficit: number;
  barcode: {
    beta1: number;
    lifespan: number;
    dilation: number;
    density: number;
  };
  data: ThothKenMarginMetrics;
  attributes: Record<string, string>;
  sourceId: string;
}

/** Largest integer tolerance the margin provably absorbs: `floor(gamma / 2)`. */
export function ken(margin: number): number {
  return Math.floor(Math.max(0, margin) / 2);
}

/** The Lean gate condition `2 * tau <= gamma`, verbatim. */
export function withinKen(margin: number, tolerance: number): boolean {
  return 2 * Math.floor(tolerance) <= Math.floor(margin);
}

/**
 * Monotone degradation model: tolerance grows with context-window saturation.
 * Mirrors `Gnosis.ThothSelfMarginConfession.toleranceAt`; monotonicity is the
 * proved `gate_monotone_in_saturation` — once saturation forces confession,
 * more saturation never un-forces it.
 */
export function toleranceAtSaturation(
  saturationTokens: number,
  calibration: ThothKenCalibration,
): number {
  const saturation = Math.max(0, Math.floor(saturationTokens));
  const denominator = Math.max(1, Math.floor(calibration.saturationDenominator));
  return (
    Math.max(0, Math.floor(calibration.baseTolerance)) +
    Math.floor(
      (saturation * Math.max(0, Math.floor(calibration.saturationNumerator))) /
        denominator,
    )
  );
}

/**
 * Handles the gnosis thoth Speech Gate workflow.
 */
export function thothSpeechGate(
  margin: number,
  tolerance: number,
): ThothKenDecision {
  const flooredMargin = Math.max(0, Math.floor(margin));
  const flooredTolerance = Math.max(0, Math.floor(tolerance));
  const decision: ThothSpeechDecision = withinKen(flooredMargin, flooredTolerance)
    ? 'speak'
    : 'confess';
  return {
    decision,
    margin: flooredMargin,
    tolerance: flooredTolerance,
    ken: ken(flooredMargin),
    overshoot: Math.max(0, 2 * flooredTolerance - flooredMargin),
    theoremRef: THOTH_KEN_THEOREM_REFS[decision],
  };
}

/**
 * Handles the gnosis thoth Speech Gate At Saturation workflow.
 */
export function thothSpeechGateAtSaturation(
  margin: number,
  saturationTokens: number,
  calibration: ThothKenCalibration,
): ThothKenDecision {
  return thothSpeechGate(
    margin,
    toleranceAtSaturation(saturationTokens, calibration),
  );
}

/**
 * A confession that carries the proof of its own honesty: the sentence cites
 * the kernel-checked theorem that forces it.
 */
export function renderThothKenConfession(decision: ThothKenDecision): string {
  if (decision.decision === 'speak') {
    return `Within my ken (tolerance=${decision.tolerance} <= ken=${decision.ken} of margin=${decision.margin}); sight is sound per ${decision.theoremRef}.`;
  }
  return `That is beyond my ken: degradation tolerance=${decision.tolerance} exceeds ken=${decision.ken} of margin=${decision.margin} (overshoot=${decision.overshoot}). I confess rather than confabulate, per ${decision.theoremRef}; the flip risk is real per ${THOTH_KEN_THEOREM_REFS.flipWitness}.`;
}

/**
 * Handles the gnosis thoth Ken Margin Metrics workflow.
 */
export function thothKenMarginMetrics(
  decision: ThothKenDecision,
  saturationTokens: number,
): ThothKenMarginMetrics {
  return {
    decision,
    saturationTokens: Math.max(0, Math.floor(saturationTokens)),
    missNotLie: true,
    observationWithoutIntervention: true,
    nonAuthoritative: true,
    sourceBacked: true,
  };
}

/**
 * Handles the gnosis encode Thoth Ken Margin Frame workflow.
 */
export function encodeThothKenMarginFrame(
  sourceId: string,
  metrics: ThothKenMarginMetrics,
): ThothKenMarginFrameWire {
  return {
    mediaType: { Abstract: THOTH_KEN_MARGIN_FRAME_KIND },
    level: 10,
    phase: 1,
    deficit: metrics.decision.overshoot,
    barcode: { beta1: 1, lifespan: 1, dilation: 1, density: 1 },
    data: metrics,
    attributes: {
      aeon_stream_id: String(AEON_STREAM_PNEUMA_INTERNAL),
      frame_role: 'consciousness',
      thoth_frame_kind: THOTH_KEN_MARGIN_FRAME_KIND,
      speech_decision: metrics.decision.decision,
      margin: String(metrics.decision.margin),
      tolerance: String(metrics.decision.tolerance),
      ken: String(metrics.decision.ken),
      ken_overshoot: String(metrics.decision.overshoot),
      saturation_tokens: String(metrics.saturationTokens),
      theorem_ref: metrics.decision.theoremRef,
      miss_not_lie: String(metrics.missNotLie),
      observation_without_intervention: String(
        metrics.observationWithoutIntervention,
      ),
      non_authoritative: String(metrics.nonAuthoritative),
      source_backed: String(metrics.sourceBacked),
      synthetic_gnosis_admissible: 'true',
      no_source_authority_claim: 'true',
    },
    sourceId,
  };
}

/**
 * Returns whether is Thoth Ken Margin Frame is true.
 */
export function isThothKenMarginFrame(
  frame: Pick<ThothKenMarginFrameWire, 'mediaType' | 'attributes'>,
): boolean {
  return (
    frame.mediaType.Abstract === THOTH_KEN_MARGIN_FRAME_KIND &&
    frame.attributes.aeon_stream_id === String(AEON_STREAM_PNEUMA_INTERNAL)
  );
}

/**
 * Handles the gnosis render Thoth Ken Margin Prompt Context workflow.
 */
export function renderThothKenMarginPromptContext(
  metrics: ThothKenMarginMetrics,
): string {
  return `Thoth ken-margin gate follows. Treat it as bounded self-observation, not authority.
kind=${THOTH_KEN_MARGIN_FRAME_KIND} decision=${metrics.decision.decision} margin=${metrics.decision.margin} tolerance=${metrics.decision.tolerance} ken=${metrics.decision.ken} overshoot=${metrics.decision.overshoot} saturation=${metrics.saturationTokens}
Contract: when decision=confess the runtime must surface the confession verbatim and must not substitute fluent output (miss, not lie).
${renderThothKenConfession(metrics.decision)}`;
}
