// @a0n/gnosis/distributed-inference-host/src/triton-stretch.ts
// triton-stretch.ts — the Triton-Stretch. Walks an
// input plane up the observer-bandwidth tower (Level 10 -> 90 -> 270),
// each tier recovering more detail than the last. The Level 90 output
// starts from the Level 10 recovery; Level 270 starts from Level 90 —
// progressive composition, not parallel ensemble. Output is a sequence of
// recovered planes, each at a higher tier of the tower. Per-level
// spectrum-band measurement reports how much new high-band content each
// tier added (the Pleromatic Stretch witness). Honest scope: this is a
// composition primitive over deblurAnyImage, not a learned multi-scale
// model. The σ-per-level values are heuristic anchors at φ-spaced grid
// points; tune via `levelSigmas` if the blur in the input is not σ ≈ 2-3.
// When `levelSigmas` is omitted, the default path runs the φ-attractor
// (`selectSigmaByGoldenAttractor`) on the mirror-padded input to derive a
// base σ, then composes the tower as [σ_base/φ, σ_base, σ_base·φ]. The
// legacy hardcoded schedule `[1.0, 2.5, 5.5]` is now opt-in via
// `useAttractor: false`.

import { deblurAnyImage, type DeblurAnyImageOptions } from './deblur-any-image';
import {
  measureSpectrumBands,
  type SpectrumBands,
} from './image-spectrum-classify';
import { mirrorPad, nextPow2 } from './image-pad';
import { selectSigmaByGoldenAttractor } from './sigma-attractor';
import { PHI } from './attractor';

export interface TritonStretchLevel {
  /** Symbolic name. */
  name: string;
  /** Tier index 0, 1, 2 corresponding to Level 10, 90, 270. */
  index: number;
  /** σ used for this level's deblur pass. */
  sigma: number;
  /** Recovered plane after this level, length width*height. */
  plane: Float32Array;
  /** Spectrum bands of the recovered plane. */
  bands: SpectrumBands;
  /** ΔhighBand vs the previous level (or vs the input for Level 10). */
  highBandDelta: number;
}

export interface TritonStretchArgs {
  /** Single-channel Float32 input. Width * height. */
  plane: Float32Array;
  width: number;
  height: number;
  /**
   * Per-tier σ. When provided, used as-is. Must have length 3.
   * When omitted, the σ schedule is auto-derived (see `useAttractor`).
   */
  levelSigmas?: readonly [number, number, number];
  /**
   * Controls the auto-derive path used when `levelSigmas` is omitted.
   * Default `true`: run `selectSigmaByGoldenAttractor` on the mirror-padded
   * input to pick a base σ, then compose [σ/φ, σ, σ·φ]. When `false`, fall
   * back to the legacy hardcoded schedule `[1.0, 2.5, 5.5]`. Ignored when
   * `levelSigmas` is provided.
   */
  useAttractor?: boolean;
  /** Forwarded to deblurAnyImage at each tier. method, noiseFloor, etc. */
  options?: DeblurAnyImageOptions;
}

export interface TritonStretchResult {
  /** Three levels, in tower-ascending order: 10, 90, 270. */
  levels: readonly [TritonStretchLevel, TritonStretchLevel, TritonStretchLevel];
  /** The final stretched plane (= levels[2].plane), for callers that just
   *  want the top-tier output without the diagnostic. */
  recovered: Float32Array;
  width: number;
  height: number;
  /**
   * Base σ chosen by the φ-attractor, when the auto-derive path ran.
   * Absent when `levelSigmas` was supplied or `useAttractor: false`.
   * Tower σ values are [attractorSigma/φ, attractorSigma, attractorSigma·φ].
   */
  attractorSigma?: number;
}

const DEFAULT_LEVEL_SIGMAS: readonly [number, number, number] = [1.0, 2.5, 5.5];

const LEVEL_NAMES: readonly [string, string, string] = [
  'Level 10 (Decagon)',
  'Level 90 (Pleroma×5/3)',
  'Level 270 (Triton triple)',
];

function isPow2(n: number): boolean {
  return Number.isInteger(n) && n > 0 && (n & (n - 1)) === 0;
}

/**
 * Measures spectrum bands on a plane, mirror-padding to pow2 dims if needed.
 * Returns the bands as observed at the (possibly padded) FFT-friendly size,
 * which keeps the band-energy comparison consistent across levels.
 */
function measureBandsPow2(
  plane: Float32Array,
  width: number,
  height: number,
): SpectrumBands {
  if (isPow2(width) && isPow2(height)) {
    return measureSpectrumBands(plane, width, height);
  }
  const padded = mirrorPad({
    plane,
    width,
    height,
    paddedWidth: nextPow2(width),
    paddedHeight: nextPow2(height),
  });
  return measureSpectrumBands(padded.plane, padded.paddedWidth, padded.paddedHeight);
}

/**
 * Handles the gnosis triton Stretch workflow.
 */
export async function tritonStretch(
  args: TritonStretchArgs,
): Promise<TritonStretchResult> {
  const { plane, width, height } = args;

  if (!Number.isInteger(width) || width <= 0) {
    throw new RangeError(
      `tritonStretch: width must be a positive integer, got ${width}`,
    );
  }
  if (!Number.isInteger(height) || height <= 0) {
    throw new RangeError(
      `tritonStretch: height must be a positive integer, got ${height}`,
    );
  }
  if (plane.length !== width * height) {
    throw new RangeError(
      `tritonStretch: plane.length (${plane.length}) != width*height (${width * height})`,
    );
  }

  // Resolve the σ schedule. Three paths:
  //   1. Caller provided `levelSigmas` → use as-is (validated below).
  //   2. `useAttractor === false` → legacy hardcoded `[1.0, 2.5, 5.5]`.
  //   3. Default → run φ-attractor on the mirror-padded input, derive
  //      [σ/φ, σ, σ·φ] from the recovered base σ.
  let resolvedSigmas: readonly [number, number, number];
  let attractorSigma: number | undefined = undefined;

  if (args.levelSigmas) {
    resolvedSigmas = args.levelSigmas;
  } else if (args.useAttractor === false) {
    resolvedSigmas = DEFAULT_LEVEL_SIGMAS;
  } else {
    // Default: derive from φ-attractor on the (mirror-padded) input.
    const padded = mirrorPad({ plane, width, height });
    const sel = selectSigmaByGoldenAttractor({
      plane: padded.plane,
      w: padded.paddedWidth,
      h: padded.paddedHeight,
    });
    attractorSigma = sel.refinedSigma;
    // The over-stretch tier (σ * φ) must produce a PSF that fits the input
    // plane: recommendPsfSize(σ) = 2·⌈3σ⌉+1, so σ_max ≤ (min(w,h) - 1) / 6.
    // This is the Princess-and-the-Pea scope check: we can only feel through
    // as many mattresses as the plane has room for.
    // recommendPsfSize(σ) = 2·⌈3σ⌉+1, so we need 2·⌈3σ⌉+1 ≤ min(w,h).
    // The safe σ-cap is (min(w,h) − 3) / 6, which guarantees ⌈3σ⌉ ≤ (min−3)/2.
    const sigmaMaxByPlaneFit = Math.max(0.5, (Math.min(args.width, args.height) - 3) / 6);
    const overstretchSigma = Math.min(attractorSigma * PHI, sigmaMaxByPlaneFit);
    const baseSigma = Math.min(attractorSigma, sigmaMaxByPlaneFit);
    const subSigma = Math.min(attractorSigma / PHI, sigmaMaxByPlaneFit);
    resolvedSigmas = [subSigma, baseSigma, overstretchSigma];
  }

  if (!Array.isArray(resolvedSigmas) && !(resolvedSigmas as readonly number[]).length) {
    throw new RangeError('tritonStretch: levelSigmas must be a tuple of length 3');
  }
  if ((resolvedSigmas as readonly number[]).length !== 3) {
    throw new RangeError(
      `tritonStretch: levelSigmas must have length 3, got ${(resolvedSigmas as readonly number[]).length}`,
    );
  }
  for (let i = 0; i < 3; i++) {
    const s = resolvedSigmas[i]!;
    if (!Number.isFinite(s) || s <= 0) {
      throw new RangeError(
        `tritonStretch: levelSigmas[${i}] must be > 0 and finite, got ${s}`,
      );
    }
  }

  const baseOptions = args.options ?? {};

  // Pre-measure the input plane's spectrum bands. Used as the prevBands for
  // Level 10's highBandDelta witness.
  const inputBands = measureBandsPow2(plane, width, height);

  const built: TritonStretchLevel[] = [];

  for (let i = 0; i < 3; i++) {
    const sourcePlane: Float32Array =
      i === 0 ? plane : built[i - 1]!.plane;

    const sigma = resolvedSigmas[i]!;

    const result = await deblurAnyImage(
      { data: sourcePlane, width, height, channels: 1 },
      { ...baseOptions, sigma },
    );

    // deblurAnyImage returns an interleaved Float32Array; for single-channel
    // it's already plane-shaped (length width*height). Make a fresh copy so
    // callers can't observe internal aliasing across levels.
    const recovered = new Float32Array(result.recovered);

    const bands = measureBandsPow2(recovered, width, height);
    const prevBands = i === 0 ? inputBands : built[i - 1]!.bands;
    const highBandDelta = bands.highBandEnergy - prevBands.highBandEnergy;

    built.push({
      name: LEVEL_NAMES[i]!,
      index: i,
      sigma,
      plane: recovered,
      bands,
      highBandDelta,
    });
  }

  const levels = [built[0]!, built[1]!, built[2]!] as const;

  return {
    levels,
    recovered: levels[2].plane,
    width,
    height,
    ...(attractorSigma !== undefined ? { attractorSigma } : {}),
  };
}
