// @a0n/gnosis/distributed-inference-host/src/tts-audio.ts
import { spawn } from 'node:child_process';
import { existsSync, mkdirSync, writeFileSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import * as voice from '@a0n/voice-tts';
import { textToPhonemeIds } from './phonemizer';

export type TtsAudioMode = 'auto' | 'host' | 'pulse' | 'alsa' | 'file';
export type ResolvedTtsAudioMode = Exclude<TtsAudioMode, 'auto'>;
export type SpeechFormat = 'wav' | 'pcm';

export interface SpeechHttpResponse {
  status: number;
  headers: Record<string, string>;
  body: Uint8Array;
}

interface ModeResolutionOptions {
  platform?: NodeJS.Platform | string;
  hasAlsaDevice?: () => boolean;
}

interface SpeechOptions extends ModeResolutionOptions {
  env?: NodeJS.ProcessEnv;
  runCommand?: (command: string, args: string[]) => Promise<boolean>;
}

const TTS_MODES = new Set<TtsAudioMode>([
  'auto',
  'host',
  'pulse',
  'alsa',
  'file',
]);
const SAMPLE_RATE = 24_000;
const PAUSE_PHONEME_ID = 1;

// The sovereign pneuma (squeezebox Klatt) voices exposed by @a0n/voice-tts.
type PneumaVoice = 'pleroma' | 'sophia' | 'aeon';

/**
 * Map OpenAI voice names and our own names onto the three squeezebox voices.
 *   pleroma | male   -> pleroma
 *   sophia  | female -> sophia
 *   everything else (alloy/echo/aeon/default/...) -> aeon
 */
function resolvePneumaVoice(name: string | undefined): PneumaVoice {
  switch (name?.trim().toLowerCase()) {
    case 'pleroma':
    case 'male':
      return 'pleroma';
    case 'sophia':
    case 'female':
      return 'sophia';
    default:
      return 'aeon';
  }
}

function floatToPcm16(samples: Float32Array): Int16Array {
  const out = new Int16Array(samples.length);
  for (let i = 0; i < samples.length; i++) {
    out[i] = Math.round(clampSample(samples[i] ?? 0) * 32767);
  }
  return out;
}

interface PneumaSpeech {
  samples: Int16Array;
  sampleRate: number;
  source: 'pneuma' | 'klatt-fallback';
}

/**
 * Synthesize speech with the sovereign pneuma voice (@a0n/voice-tts). Falls back
 * to the legacy Klatt synth only if voice-tts throws.
 */
export function synthesizePneumaSpeech(
  input: string,
  voiceName?: string,
): PneumaSpeech {
  const pneumaVoice = resolvePneumaVoice(voiceName);
  try {
    const pcm = voice.synthesize(input, pneumaVoice);
    if (!pcm || pcm.length === 0) {
      throw new Error('voice-tts returned empty PCM');
    }
    return {
      samples: floatToPcm16(pcm),
      sampleRate: voice.sample_rate(),
      source: 'pneuma',
    };
  } catch {
    return {
      samples: synthesizeSpeechPcm16(input, voiceName),
      sampleRate: SAMPLE_RATE,
      source: 'klatt-fallback',
    };
  }
}

interface Qwen3TtsSpeech {
  samples: Int16Array;
  sampleRate: number;
  source: 'qwen3-tts';
}

/**
 * End-to-end sovereign neural TTS via the Qwen3-TTS talker + carried codec
 * decoder (forkjoin-ai/qwen3-tts-12hz-0.6b-base), run through the native
 * `tts-probe --talker` binary against the talker-enabled knot.
 *
 * Gated behind `MOONSHINE_TTS_QWEN3=1` (Klatt / pneuma remain the default and
 * the fallback). Requires:
 *   MOONSHINE_TTS_QWEN3_BIN   path to the tts-probe binary (fat-station family)
 *   MOONSHINE_TTS_QWEN3_KNOT  path/URL to the talker-enabled .knot
 * Tokenization (text -> chat-templated Qwen text-token ids) is delegated to the
 * tts-probe binary's `--text` path; until that lands, the host can pass
 * pre-tokenized ids via `MOONSHINE_TTS_QWEN3_TEXTIDS_CMD` (a command that prints
 * a comma-separated id list for the input on stdin). Returns null on any failure
 * so the caller can fall back to pneuma/Klatt without surfacing a hard error.
 */
export async function synthesizeQwen3TtsSpeech(
  input: string,
  env: NodeJS.ProcessEnv,
  options: {
    runCapture?: (
      command: string,
      args: string[],
      stdin?: string,
    ) => Promise<{ ok: boolean; stdout: string }>;
  } = {},
): Promise<Qwen3TtsSpeech | null> {
  if (env['MOONSHINE_TTS_QWEN3'] !== '1') return null;
  const bin = env['MOONSHINE_TTS_QWEN3_BIN'];
  const knot = env['MOONSHINE_TTS_QWEN3_KNOT'];
  if (!bin || !knot) return null;
  const runCapture = options.runCapture ?? defaultRunCapture;

  try {
    // 1. Resolve chat-templated text-token ids for the input.
    let textIds: string | null = null;
    const tokCmd = env['MOONSHINE_TTS_QWEN3_TEXTIDS_CMD'];
    if (tokCmd) {
      const [cmd, ...cargs] = tokCmd.split(' ');
      const r = await runCapture(cmd!, cargs, input);
      if (r.ok && r.stdout.trim().length > 0) textIds = r.stdout.trim();
    }

    // 2. Run the talker (writes a WAV; prints structural-gate result).
    const outDir = env['MOONSHINE_TTS_OUTPUT_DIR'] ?? join(tmpdir(), 'moonshine-tts');
    mkdirSync(outDir, { recursive: true });
    const outWav = join(outDir, `qwen3-${Date.now()}.wav`);
    const args = ['--knot', knot, '--out', outWav];
    if (textIds) {
      args.push('--text-ids', textIds);
    } else {
      // Binary-side tokenization (tts-probe --text), once available.
      args.push('--text', input);
    }
    const run = await runCapture(bin, args);
    if (!run.ok || !existsSync(outWav)) return null;

    // 3. Read the WAV back into PCM16 (mono, 24 kHz).
    const { readFileSync } = await import('node:fs');
    const wav = readFileSync(outWav);
    const parsed = parseWavPcm16(new Uint8Array(wav));
    if (!parsed || parsed.samples.length === 0) return null;
    return {
      samples: parsed.samples,
      sampleRate: parsed.sampleRate,
      source: 'qwen3-tts',
    };
  } catch {
    return null;
  }
}

/** Parse a canonical PCM16 mono WAV (the layout tts-probe writes). */
function parseWavPcm16(
  bytes: Uint8Array,
): { samples: Int16Array; sampleRate: number } | null {
  if (bytes.length < 44) return null;
  const dv = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  // RIFF....WAVE
  if (dv.getUint32(0, false) !== 0x52494646) return null; // 'RIFF'
  if (dv.getUint32(8, false) !== 0x57415645) return null; // 'WAVE'
  const sampleRate = dv.getUint32(24, true);
  // data chunk: scan for 'data'
  let p = 12;
  while (p + 8 <= bytes.length) {
    const id = dv.getUint32(p, false);
    const size = dv.getUint32(p + 4, true);
    if (id === 0x64617461) {
      // 'data'
      const n = Math.min(size, bytes.length - (p + 8)) >> 1;
      const samples = new Int16Array(n);
      for (let i = 0; i < n; i++) samples[i] = dv.getInt16(p + 8 + i * 2, true);
      return { samples, sampleRate };
    }
    p += 8 + size + (size & 1);
  }
  return null;
}

function defaultRunCapture(
  command: string,
  args: string[],
  stdin?: string,
): Promise<{ ok: boolean; stdout: string }> {
  return new Promise((resolve) => {
    const child = spawn(command, args, { stdio: ['pipe', 'pipe', 'ignore'] });
    let stdout = '';
    const timer = setTimeout(() => child.kill('SIGTERM'), 120_000);
    child.stdout?.on('data', (d) => {
      stdout += d.toString();
    });
    child.on('close', (code) => {
      clearTimeout(timer);
      resolve({ ok: code === 0, stdout });
    });
    child.on('error', () => {
      clearTimeout(timer);
      resolve({ ok: false, stdout });
    });
    if (stdin !== undefined) {
      child.stdin?.write(stdin);
      child.stdin?.end();
    }
  });
}

/**
 * Word timings (karaoke) for the given input, as a JSON string, or undefined if
 * voice-tts cannot produce a frame. Drawn from voice_frame's `word_timings`.
 */
export function pneumaWordTimings(
  input: string,
  voiceName?: string,
): string | undefined {
  try {
    const frame = JSON.parse(voice.voice_frame(input, resolvePneumaVoice(voiceName)));
    const timings = (frame as { word_timings?: unknown })?.word_timings;
    return timings === undefined ? undefined : JSON.stringify(timings);
  } catch {
    return undefined;
  }
}

const PHONEME_FORMANTS: ReadonlyArray<readonly [number, number]> = [
  [0, 0],
  [0, 0],
  [190, 760],
  [700, 1220],
  [660, 1720],
  [540, 1280],
  [570, 840],
  [700, 1120],
  [710, 1320],
  [530, 1840],
  [500, 1480],
  [360, 2220],
  [320, 2480],
  [470, 900],
  [530, 1020],
  [350, 860],
  [390, 1640],
  [410, 1460],
  [330, 960],
  [240, 760],
  [300, 1900],
  [260, 1180],
  [230, 980],
  [310, 1780],
  [260, 860],
  [180, 820],
  [290, 1420],
  [280, 1740],
  [260, 760],
  [260, 900],
  [220, 1040],
  [210, 980],
  [250, 900],
  [210, 730],
  [280, 1500],
  [320, 2100],
  [350, 2600],
  [280, 900],
  [340, 2100],
  [270, 880],
  [300, 920],
  [310, 1360],
  [330, 1840],
];

function jsonResponse(status: number, payload: unknown): SpeechHttpResponse {
  return {
    status,
    headers: { 'Content-Type': 'application/json' },
    body: new TextEncoder().encode(JSON.stringify(payload)),
  };
}

/**
 * Parses the Tts Audio Mode.
 */
export function parseTtsAudioMode(value: string | undefined): TtsAudioMode {
  const normalized = value?.trim().toLowerCase();
  if (!normalized) return 'auto';
  return TTS_MODES.has(normalized as TtsAudioMode)
    ? (normalized as TtsAudioMode)
    : 'auto';
}

/**
 * Resolves the Tts Audio Mode.
 */
export function resolveTtsAudioMode(
  value: string | undefined,
  options: ModeResolutionOptions = {},
): ResolvedTtsAudioMode {
  const requested = parseTtsAudioMode(value);
  if (requested !== 'auto') return requested;

  const platform = options.platform ?? process.platform;
  if (platform === 'darwin') return 'host';
  if (
    platform === 'linux' &&
    (options.hasAlsaDevice ?? (() => existsSync('/dev/snd')))()
  ) {
    return 'alsa';
  }
  return 'file';
}

function writeAscii(view: DataView, offset: number, value: string): void {
  for (let i = 0; i < value.length; i++) {
    view.setUint8(offset + i, value.charCodeAt(i));
  }
}

function clampSample(sample: number): number {
  return Math.max(-1, Math.min(1, sample));
}

function voiceOffset(voice: string | undefined): number {
  if (!voice) return 0;
  let hash = 0;
  for (let i = 0; i < voice.length; i++) {
    hash = (hash * 31 + voice.charCodeAt(i)) >>> 0;
  }
  return (hash % 45) - 22;
}

function formantsForPhoneme(phonemeId: number): readonly [number, number] {
  return (
    PHONEME_FORMANTS[phonemeId] ?? [
      240 + (phonemeId % 17) * 18,
      900 + (phonemeId % 29) * 47,
    ]
  );
}

function isPausePhoneme(phonemeId: number): boolean {
  return phonemeId === PAUSE_PHONEME_ID || phonemeId <= 0;
}

/**
 * Handles the gnosis synthesize Speech Pcm16 workflow.
 */
export function synthesizeSpeechPcm16(input: string, voice?: string): Int16Array {
  const text = input.trim();
  const phonemes = textToPhonemeIds(text);
  const symbols = phonemes.length > 0 ? phonemes : [PAUSE_PHONEME_ID];
  const durationSeconds = Math.max(
    0.45,
    Math.min(8, 0.08 + symbols.length * 0.075),
  );
  const sampleCount = Math.max(1, Math.floor(durationSeconds * SAMPLE_RATE));
  const samples = new Int16Array(sampleCount);
  const symbolSpan = sampleCount / symbols.length;
  const offset = voiceOffset(voice);

  for (let i = 0; i < sampleCount; i++) {
    const t = i / SAMPLE_RATE;
    const phonemeIndex = Math.min(
      symbols.length - 1,
      Math.floor(i / symbolSpan),
    );
    const phonemeId = symbols[phonemeIndex]!;
    const voiced = !isPausePhoneme(phonemeId);
    const [baseFormant, upperFormant] = formantsForPhoneme(phonemeId);
    const base = voiced ? baseFormant + offset : 0;
    const formant = voiced ? upperFormant + offset * 2 : 0;
    const localPhase = (i - phonemeIndex * symbolSpan) / symbolSpan;
    const attack = Math.min(1, localPhase * 8);
    const release = Math.min(1, (1 - localPhase) * 6);
    const envelope = voiced ? 0.25 * Math.min(attack, release) : 0;
    const sample =
      Math.sin(2 * Math.PI * base * t) * envelope +
      Math.sin(2 * Math.PI * formant * t) * envelope * 0.28 +
      Math.sin(2 * Math.PI * (base + formant) * 0.5 * t) * envelope * 0.16;
    samples[i] = Math.round(clampSample(sample) * 32767);
  }

  return samples;
}

/**
 * Handles the gnosis encode Wav Pcm16 workflow.
 */
export function encodeWavPcm16(
  samples: Int16Array,
  sampleRate = SAMPLE_RATE,
): Uint8Array {
  const dataBytes = samples.length * 2;
  const buffer = new ArrayBuffer(44 + dataBytes);
  const view = new DataView(buffer);
  writeAscii(view, 0, 'RIFF');
  view.setUint32(4, 36 + dataBytes, true);
  writeAscii(view, 8, 'WAVE');
  writeAscii(view, 12, 'fmt ');
  view.setUint32(16, 16, true);
  view.setUint16(20, 1, true);
  view.setUint16(22, 1, true);
  view.setUint32(24, sampleRate, true);
  view.setUint32(28, sampleRate * 2, true);
  view.setUint16(32, 2, true);
  view.setUint16(34, 16, true);
  writeAscii(view, 36, 'data');
  view.setUint32(40, dataBytes, true);
  for (let i = 0; i < samples.length; i++) {
    view.setInt16(44 + i * 2, samples[i]!, true);
  }
  return new Uint8Array(buffer);
}

function pcmBytes(samples: Int16Array): Uint8Array {
  return new Uint8Array(samples.buffer, samples.byteOffset, samples.byteLength);
}

function writeTtsFile(bytes: Uint8Array, env: NodeJS.ProcessEnv): string {
  const outputDir =
    env['MOONSHINE_TTS_OUTPUT_DIR'] ?? join(tmpdir(), 'moonshine-tts');
  mkdirSync(outputDir, { recursive: true });
  const path = join(
    outputDir,
    `speech-${Date.now()}-${Math.random().toString(36).slice(2)}.wav`,
  );
  writeFileSync(path, bytes);
  return path;
}

function defaultRunCommand(command: string, args: string[]): Promise<boolean> {
  return new Promise((resolve) => {
    const child = spawn(command, args, { stdio: 'ignore' });
    const timer = setTimeout(() => {
      child.kill('SIGTERM');
    }, 15_000);
    child.on('close', (code) => {
      clearTimeout(timer);
      resolve(code === 0);
    });
    child.on('error', () => {
      clearTimeout(timer);
      resolve(false);
    });
  });
}

async function maybePlayInContainer(
  mode: ResolvedTtsAudioMode,
  wavBytes: Uint8Array,
  env: NodeJS.ProcessEnv,
  runCommand: (command: string, args: string[]) => Promise<boolean>,
): Promise<{ playback: string; filePath?: string }> {
  if (mode === 'host') {
    return { playback: 'host-relay' };
  }

  if (mode === 'file') {
    return { playback: 'file', filePath: writeTtsFile(wavBytes, env) };
  }

  const filePath = writeTtsFile(wavBytes, env);
  const played =
    mode === 'pulse'
      ? await runCommand('paplay', [filePath])
      : await runCommand('aplay', [filePath]);
  return {
    playback: played ? mode : `${mode}-failed`,
    filePath,
  };
}

/**
 * Handles the Speech Request Body request flow.
 */
export async function handleSpeechRequestBody(
  body: string,
  options: SpeechOptions = {},
): Promise<SpeechHttpResponse> {
  let parsed: unknown;
  try {
    parsed = JSON.parse(body);
  } catch {
    return jsonResponse(400, {
      error: { message: 'invalid JSON', type: 'invalid_request_error' },
    });
  }

  const request = parsed as {
    input?: unknown;
    voice?: unknown;
    format?: unknown;
  };
  if (typeof request.input !== 'string' || request.input.trim().length === 0) {
    return jsonResponse(400, {
      error: { message: 'input must be a non-empty string', type: 'invalid_request_error' },
    });
  }
  if (request.voice !== undefined && typeof request.voice !== 'string') {
    return jsonResponse(400, {
      error: { message: 'voice must be a string when provided', type: 'invalid_request_error' },
    });
  }

  const format = (request.format ?? 'wav') as SpeechFormat;
  if (format !== 'wav' && format !== 'pcm') {
    return jsonResponse(400, {
      error: { message: 'format must be "wav" or "pcm"', type: 'invalid_request_error' },
    });
  }

  const env = options.env ?? process.env;
  // Sovereign neural TTS (Qwen3-TTS talker + codec), gated by MOONSHINE_TTS_QWEN3.
  // Klatt / pneuma stays the default and the fallback on any failure.
  const qwen3 = await synthesizeQwen3TtsSpeech(request.input, env);
  const speech = qwen3 ?? synthesizePneumaSpeech(request.input, request.voice);
  const samples = speech.samples;
  const wavBytes = encodeWavPcm16(samples, speech.sampleRate);
  const mode = resolveTtsAudioMode(env['MOONSHINE_TTS_AUDIO_MODE'], options);
  const playback = await maybePlayInContainer(
    mode,
    wavBytes,
    env,
    options.runCommand ?? defaultRunCommand,
  );
  const bodyBytes = format === 'wav' ? wavBytes : new Uint8Array(pcmBytes(samples));
  const wordTimings =
    speech.source === 'pneuma'
      ? pneumaWordTimings(request.input, request.voice)
      : undefined;

  return {
    status: 200,
    headers: {
      'Content-Type': format === 'wav' ? 'audio/wav' : 'application/octet-stream',
      'X-Moonshine-TTS-Audio-Mode': mode,
      'X-Moonshine-TTS-Playback': playback.playback,
      'X-Moonshine-TTS-Synth': speech.source,
      ...(playback.filePath ? { 'X-Moonshine-TTS-File': playback.filePath } : {}),
      ...(wordTimings ? { 'X-Pneuma-Word-Timings': wordTimings } : {}),
    },
    body: bodyBytes,
  };
}
