// @a0n/gnosis/distributed-inference-host/src/types.ts
/**
 * Public types for @a0n/distributed-inference-host.
 *
 * The host shim talks to two collaborators:
 *   1. A KnotByteSource — a transport that resolves byte ranges from a knot
 *      file. R2 service bindings, raw HTTPS, file system, and in-memory all
 *      satisfy this single interface so the shim itself is transport-agnostic.
 *   2. A WasmModule — the JS-callable surface produced by wasm-bindgen for
 *      open-source/gnosis/distributed-inference (wasm32 build).
 */

import type { BuleyProjection } from './spectral-noise-monitor';

/**
 * Transport interface every Worker deployment provides. Range bounds are
 * inclusive on both ends to mirror the HTTP `Range: bytes=start-end` header
 * and Cloudflare R2's `range: { offset, length }` semantics after conversion.
 */
export interface KnotByteSource {
  /**
   * Fetch bytes [start, end] inclusive from the knot file.
   *
   * Implementations may serve a longer slice when the upstream caches at a
   * coarser granularity, but the returned Uint8Array MUST cover the requested
   * inclusive range starting at `start`.
   */
  rangeGet(start: number, end: number): Promise<Uint8Array>;
}

/** Raw tensor descriptor as emitted by the Rust crate's metadata. */
export interface TensorInfo {
  offset: number;
  length: number;
  /** GGUF-style quant tag, e.g. "Q4_K", "F16", "F32". */
  tensor_type: string;
}

/**
 * The subset of WasmHostBackend we depend on. Defined structurally so callers
 * can pass either the real wasm-bindgen class or a mock without coupling to
 * the wasm module's exact constructor identity.
 */
export interface WasmHostBackendLike {
  set_tensor(name: string, bytes: Uint8Array): void;
  evict(names: string[]): void;
  tensor_info(name: string): TensorInfo | null;
}

/** The subset of WasmLlamaPipeline we depend on. */
export interface WasmLlamaPipelineLike {
  generate(prompt_tokens: Uint32Array, max_new_tokens: number): Uint32Array;
  get_pos(): number;
  /** TQFT Cobordism: vocabulary-sharded embed. Optional — only present
   *  when the WASM module exposes embed_range (requires crate rebuild). */
  embed_range?(
    tokens: Uint32Array,
    start_pos: number,
    vocab_start: number,
    vocab_end: number
  ): Float32Array;
  /** TQFT Cobordism: vocabulary-sharded lm_head. Optional. */
  lm_head_range?(
    residual: Float32Array,
    sequence_length: number,
    vocab_start: number,
    vocab_end: number
  ): Float32Array;
}

/**
 * Constructor + static surface the host shim consumes from the wasm module.
 * Matches the wasm-bindgen output of the distributed-inference crate.
 */
export interface WasmModule {
  WasmHostBackend: new (metadata_json: string) => WasmHostBackendLike;
  WasmLlamaPipeline: {
    from_backend(
      backend: WasmHostBackendLike,
      config_json: string | null,
      payload_offset: number,
      draft_mode: boolean,
      resonance_threshold: number
    ): WasmLlamaPipelineLike;
  };
  /** Optional helpers the wasm crate exposes for parsing metadata in-WASM. */
  parse_tensors_json?: (metadata_json: string) => string;
  parse_config_json?: (metadata_json: string) => string | null;
}

export interface AckermannAdmissionGate {
  projection: BuleyProjection;
  replicas: number;
  ceiling?: number;
}

/** Knot metadata JSON shape. */
export interface KnotMetadata {
  config?: Record<string, unknown> | null;
  tensors: Record<string, TensorInfo>;
}

/**
 * IStation — Transport-agnostic interface for an inference station.
 * Satisfied by both `Station` (HTTP/Fetch) and `StationUds` (Unix Domain Socket).
 */
/**
 * Orthographic style preferences for personality logit-steering (signed:
 * negative = an anti-tell the contact avoids, e.g. emoji for an emoji-hater).
 * Mirrors `personality_steering::StylePreferences` on the station.
 */
export interface PersonalityStylePreferences {
  caps?: number;
  lower?: number;
  exclaim?: number;
  question?: number;
  ellipsis?: number;
  period?: number;
  emoji?: number;
  ascii_emoticon?: number;
}

/** Mirrors `personality_steering::SteeringSignature` (four wave-axis gains + caps). */
export interface PersonalitySteeringSignature {
  behavioral_gain?: number;
  aesthetic_gain?: number;
  prosody_gain?: number;
  harmonic_gain?: number;
  steer_cap?: number;
  rest_cap?: number;
}

/** Install form for `POST /personality-steering`. */
export interface PersonalitySteeringSpec {
  stylePreferences: PersonalityStylePreferences;
  signature: PersonalitySteeringSignature;
  /** Off-boundary envelope scale (≈ rest_cap / steer_cap). Defaults station-side. */
  restScale?: number;
  /**
   * Precompiled per-token signed bias vector (Path A). When present, it is sent
   * to the station verbatim instead of asking it to compile from
   * stylePreferences — required when the station's tokenizer is not a GGUF the
   * fat-station can read (e.g. mistral's SentencePiece/JSON tokenizer). The
   * openai-server compiles this against its own tokenizer vocab.
   */
  boundaryBias?: number[];
  /**
   * SIGNED conflict-register gains from the abductive spectral model (twin-forge
   * `SignedGains`). Unlike `signature` (the unsigned orthographic envelope), these
   * promote/demote semantic register tokens. The openai-server resolves them to a
   * per-token bias against its own vocab (`conflict-register.compileSemanticBias`)
   * and folds the result into `boundaryBias`. Steers WHAT a contact says under
   * conflict (cold/curt vs warm/repair), not just HOW they punctuate it.
   */
  semantic?: {
    behavioral: number;
    aesthetic: number;
    prosody: number;
    harmonic: number;
    cap: number;
  };
}

export interface IStation {
  readonly startLayer: number;
  readonly endLayer: number;
  readonly role?: 'entry' | 'mid' | 'exit' | 'cobordism';
  /**
   * Per-station batch capacity advertised to Pipeline's chunkSize clamp.
   * The Rust kernel caps batched-token state at `BATCH_CAPACITY` per call:
   *   wasm32 target:  4   (apps/distributed-inference-worker, *.workers.dev)
   *   native target:  8   (fat-station local mesh, bench harnesses)
   * (see open-source/gnosis/distributed-inference/src/model.rs:381-383).
   *
   * Optional so existing implementations satisfy the structural contract;
   * Pipeline takes the min over all stations and falls back to the wasm-safe
   * `4` when a station omits this field.
   */
  readonly maxBatchSize?: number;

  forward(args: {
    residual: Float32Array;
    startPos: number;
    sequenceLength: number;
    requestId?: string;
  }): Promise<{
    residual: Float32Array;
    teleportation?: {
      layerDeficits: number[];
      rounds: number;
    };
  }>;

  embed(args: {
    tokens: Uint32Array;
    startPos: number;
    requestId?: string;
  }): Promise<Float32Array>;

  lmHead(args: {
    residual: Float32Array;
    sequenceLength: number;
    requestId?: string;
    ackermannAdmission?: AckermannAdmissionGate;
  }): Promise<Float32Array>;

  sampleNext?(args: {
    residual: Float32Array;
    sequenceLength: number;
    requestId?: string;
    ackermannAdmission?: AckermannAdmissionGate;
  }): Promise<number>;

  /** Exit-only: install a personality logit-steering filter (opt-in). */
  installPersonalitySteering?(
    spec: PersonalitySteeringSpec,
    requestId?: string
  ): Promise<void>;

  /** Exit-only: clear any installed personality steering (restore base model). */
  clearPersonalitySteering?(requestId?: string): Promise<void>;

  decodeNext?(args: {
    tokenId: number;
    position: number;
    requestId?: string;
    ackermannAdmission?: AckermannAdmissionGate;
    terminalProsody?: {
      emittedTokens: number;
      maxTokens: number;
      tailState: 'open-clause' | 'sentence-closed';
      sentenceCloseTokenIds?: readonly number[];
      stopTokenIds?: readonly number[];
    };
  }): Promise<number>;

  generateTokens?(args: {
    tokens: Uint32Array;
    maxNewTokens: number;
    requestId?: string;
    terminalProsody?: {
      sentenceCloseTokenIds?: readonly number[];
      stopTokenIds?: readonly number[];
      tailState?: 'open-clause' | 'sentence-closed';
    };
  }): Promise<Uint32Array>;

  /**
   * Sovereign-sieve FOLD accessor. Same native `/generate` as `generateTokens`,
   * but carries `return_hidden: true` so ONE forward returns BOTH the generated
   * tokens AND `promptHidden` — the last prompt token's post-all-layers tail
   * residual (pre-lm_head, shape `[hiddenDim]`), the validated semantic-sieve key.
   * See fat-station.rs `return_hidden`/`prompt_hidden` (commit 02618367). Returns
   * `promptHidden: null` when the station serves generation without surfacing the
   * hidden (capability miss → caller falls back to `computePromptHidden` or skips
   * the sieve). `return_hidden` forces the fat-station's SERIAL decode path (the
   * continuous-batch scheduler does not surface per-request hidden state), so this
   * is used only for the short agreement-gate probe, never the full decode.
   */
  generateTokensWithHidden?(args: {
    tokens: Uint32Array;
    maxNewTokens: number;
    requestId?: string;
  }): Promise<{ tokens: Uint32Array; promptHidden: Float32Array | null }>;

  /**
   * Prompt-lookup / n-gram speculative-decode VERIFY primitive (greedy,
   * LOSSLESS). Forwards the `candidates` span in ONE batched forward, accepts
   * the longest argmax-matching prefix, and commits `accepted ++ [nextToken]`
   * to the station KV. The station's KV/pos is assumed to already sit at the end
   * of the running sequence (i.e. immediately after a prefill or a prior
   * verify/decode step), with its logits holding the next-token distribution for
   * the FIRST candidate's slot.
   *
   * Returns the accepted candidate prefix, the corrected/continuation token
   * `nextToken` (always emitted — a correction when accepted < candidates, a
   * continuation when all accepted), and the resulting absolute `pos`
   * (== positionBefore + accepted.length + 1, which equals the host's running
   * seq length after it appends `accepted ++ [nextToken]`).
   */
  verifyBatch?(args: {
    candidates: Uint32Array | readonly number[];
    /**
     * Absolute KV slot index of the FIRST candidate (== the host's running
     * `seq.length` this round). Sent as `X-Position` and used by the fat-station
     * to seed `self.pos` BEFORE the batched candidate forward, so the warm prompt
     * KV is never clobbered. Every other decode route (`/decode-next`, `/forward`)
     * is likewise driven by an explicit `X-Position`; `/verify-batch` previously
     * read the station's internal `self.pos`, which the host's `/forward` prefill
     * path never sets (stale 0 → candidate KV written at slot 0 → garbage). The
     * host MUST pass this on every call.
     */
    position?: number;
    /** Optional re-prefill prompt (normally EMPTY — the prompt is already prefilled). */
    promptTokens?: Uint32Array | readonly number[];
    requestId?: string;
  }): Promise<{
    accepted: number[];
    nextToken: number;
    acceptedCount: number;
    candidateCount: number;
    pos: number;
  }>;

  /**
   * Resume-anywhere KV-cache checkpoint for the uniform-KV transformer archs:
   * the native Llama family (/health `arch` "transformer") AND Gemma
   * (/health `arch` "gemma") — both expose `get_kv_slab`/`set_kv_slab` over a
   * single `kv_cache::KVCache`. Gate on the /health `kv_session` flag (true for
   * transformer | gemma). SSM/hybrid archs (mamba/rwkv7/hybrid/qwen35moe) have
   * no uniform KV cache and return HTTP 400 here — they resume via
   * {@link saveSsmState}/{@link loadSsmState} instead. Returns false on 404.
   */
  saveKvSession?(args: {
    sessionId: string;
    prefixLength: number;
    requestId?: string;
  }): Promise<boolean>;

  loadKvSession?(args: {
    sessionId: string;
    requestId?: string;
  }): Promise<{ prefixLength: number } | null>;

  deleteKvSession?(args: {
    sessionId: string;
    requestId?: string;
  }): Promise<boolean>;

  /**
   * O(1) resume-anywhere SSM checkpoint (RWKV-7 + Mamba / FalconMamba), plus the
   * MIXED-checkpoint hybrids (Jamba/Falcon-H1 "hybrid" and Qwen3.5-MoE /
   * Qwen3-Next "qwen35moe" Gated-DeltaNet).
   * Serializes the pipeline's recurrent state — RWKV-7: per layer WKV matrix +
   * token-shift + channel-mix + v_first; Mamba: per layer causal-conv1d buffer +
   * SSM hidden state; hybrid/qwen35moe: O(1) recurrent state for the SSM layers +
   * a KV slab for the sparse attention layers — and returns it as raw bytes. For
   * the pure-SSM archs the blob is **constant-size regardless of how many tokens
   * have been generated**. Only fat-stations serving an SSM/hybrid knot (arch
   * rwkv7 | mamba | falcon-mamba | hybrid | qwen35moe) expose /ssm-state/* (gate
   * on the /health `ssm_state` or `alt_arch` flag). Returns null when the route
   * is absent (404).
   */
  saveSsmState?(args: {
    requestId?: string;
  }): Promise<Uint8Array | null>;

  /**
   * Restore a checkpoint produced by {@link saveSsmState} into the live
   * pipeline state. Restoring into a FRESH fat-station serving the SAME knot
   * reproduces the exact continuation (resume-anywhere). Returns the restored
   * byte count + layer count, or null when the route is absent (404).
   */
  loadSsmState?(args: {
    state: Uint8Array | ArrayBuffer;
    requestId?: string;
  }): Promise<{ restoredBytes: number; numLayers: number } | null>;

  cobordismEmbed(args: {
    tokens: Uint32Array;
    startPos: number;
    vocabStart: number;
    vocabEnd: number;
    requestId?: string;
  }): Promise<Float32Array>;

  cobordismLmHead(args: {
    residual: Float32Array;
    sequenceLength: number;
    vocabStart: number;
    vocabEnd: number;
    requestId?: string;
  }): Promise<Float32Array>;

  bizarroForward(args: {
    residual: Float32Array;
    startPos: number;
    sequenceLength: number;
    targetPosition: number;
    requestId?: string;
  }): Promise<{ rejectMask: number[] } | null>;

  /**
   * Amplituhedron volume-cache replay. The host coordinator
   * (Pipeline.tryAmplituhedronReplay → pipeline.ts:865) fans this out across
   * every station and SKIPS prefill on an ALL-hit. Mid-pipeline stations
   * participate by reseeding their internal KV slabs (the load-bearing side
   * effect); only the EXIT station is required to populate `tailResidual`,
   * which is then fed into lm_head.
   *
   * Optional — stations that omit this are treated as permanent misses and
   * the coordinator silently falls back to a normal prefill. See
   * AMPLITUHEDRON_PLAN.md §4 for the wire protocol.
   */
  amplituhedronReplay?(args: {
    prefixHash: string;
    prefixLen: number;
    layerLo: number;
    layerHi: number;
    requestId?: string;
  }): Promise<{
    hit: boolean;
    tailResidual?: Float32Array;
    tailFingerprint?: string;
    prefixResidual?: Float32Array;
    prefixFingerprint?: string;
  }>;

  /**
   * Fire-and-forget capture into the Amplituhedron volume cache. Called by
   * the coordinator after a prefill MISS so the next same-prefix call can
   * replay. The coordinator NEVER awaits this (scheduleAmplituhedronCapture
   * → pipeline.ts:964 uses `fireAndForget`) — implementations MUST NOT
   * throw on the decode-loop critical path. Defer to `ctx.waitUntil` or
   * equivalent so capture cost stays off the hot path.
   */
  amplituhedronCapture?(args: {
    prefixHash: string;
    prefixLen: number;
    layerLo: number;
    layerHi: number;
    tailResidual?: Float32Array;
    requestId?: string;
  }): Promise<void>;
}

/** Construction options for HostShim.open. */
export interface HostShimOptions {
  /**
   * How many bytes to fetch on the initial header probe. Must be large enough
   * to cover the magic + length prefix + the entire metadata JSON + the codec
   * marker. 4 MiB is generous for sub-100B models; larger metadata can bump
   * this. The shim re-fetches only if the declared metadata length spills
   * past the initial window.
   */
  headerFetchBytes?: number;
  /**
   * Maximum number of transformer layers (decoder blocks) held resident at
   * once. The LRU evicts the least-recently prefetched layer when this is
   * exceeded. Residents (token_embd, output_norm, output) do NOT count.
   */
  lruCapacity?: number;
  draftMode?: boolean;
  resonanceThreshold?: number;
}
