// @a0n/gnosis/distributed-inference-host/src/ultrasonic-proximity.ts
/**
 * # Ultrasonic Proximity Graph — reciprocal edges only, spoof-beacon refutation
 *
 * Mirrors `Gnosis.Mesh.UltrasonicProximityGraph` from `open-source/gnosis-math`,
 * the Rustic Church carrier for the Monster Mesh ultrasonic proximity surface.
 * Mesh devices emit inaudible ultrasonic chirps and listen for each other; from
 * the round-trip time-of-flight (ToF) of a chirp a device estimates how far away
 * its neighbour is, and from these pairwise ranges the mesh builds a proximity
 * graph. An edge "device A is near device B" exists only when the ranging is
 * physically self-consistent. The Skeptic admits a proximity EDGE only when the
 * measurement is consistent with two real devices actually ranging each other,
 * refuting a **spoofed beacon** — a device that simply announces "I'm near X"
 * without producing the reciprocal acoustic arrival a real neighbour would.
 *
 * ## What is (and is not) carried
 *
 * This module is the runtime promotion check for a proximity-edge observation.
 * It carries ONLY the finite contract the Lean file proves (one-way ToFs modeled
 * as `Nat`, speed normalized to one distance-unit per tick):
 *
 *  - the saturating ToF residual metric (`adist`);
 *  - the reciprocity gate (the core hard negative): a real edge has symmetric
 *    ToF, A->B and B->A agreeing up to tolerance (`reciprocityResidual` /
 *    `reciprocal`);
 *  - the support / non-triviality gate (source-only analog): a claim with zero
 *    measured ToF in both directions carries no ranging information
 *    (`trivialClaim` / `nonTrivial`);
 *  - edge admission: reciprocal AND non-trivial (`admitProximityEdge`);
 *  - the triangle-consistency hard negative via a third ear
 *    (`triangleConsistent`).
 *
 * Real ToF estimation, clock drift, multipath, ultrasonic channel modeling, and
 * Euclidean geometry in 2-D/3-D stay runtime/empirical obligations OUTSIDE this
 * certificate — they are not theorems and are not asserted here. (Here distance
 * is the 1-D corridor metric `adist`, which equals Euclidean distance on a line.)
 *
 * The promotion check folds the contract into the same epistemic-outcome
 * vocabulary the sibling mesh carriers already speak
 * (`claim_admitted` / `observation_only` / `claim_withheld`), so a proximity
 * edge cannot be displayed until it has cleared the Lean-side
 * reciprocity/support gate.
 *
 * ## Lean source-of-truth
 *
 * @see open-source/gnosis-math/Gnosis/Mesh/UltrasonicProximityGraph.lean
 * @see open-source/gnosis-math/Gnosis/Mesh/CrossNodeCoincidence.lean (mesh sibling)
 */

/* ------------------------------------------------------------------ */
/*  ToF distance and proximity claims                                 */
/* ------------------------------------------------------------------ */

/** Absolute difference on `Nat` via saturating subtraction (`|a - b|`). Used
 *  both as the 1-D corridor distance and as the reciprocity residual between the
 *  two directional ToFs. Mirrors `UltrasonicProximityGraph.adist`. */
export function adist(a: number, b: number): number {
  return Math.max(0, a - b) + Math.max(0, b - a);
}

/**
 * A reported proximity edge between two device indices. The device measures the
 * one-way ToF of its chirp to the neighbour (`aToB`) and of the neighbour's
 * chirp back (`bToA`), and asserts a distance estimate (`claimedDist`). A genuine
 * edge has `aToB ~ bToA` (reciprocity) and at least one nonzero arrival
 * (support). Mirrors `UltrasonicProximityGraph.ProximityClaim`.
 */
export interface ProximityClaim {
  readonly aToB: number;
  readonly bToA: number;
  readonly claimedDist: number;
}

/* ------------------------------------------------------------------ */
/*  Gate 1: reciprocity                                               */
/* ------------------------------------------------------------------ */

/** Reciprocity residual: how asymmetric the two directional ToFs are. A real
 *  edge between two co-present devices has a small residual; a one-sided spoof
 *  beacon does not. Mirrors `UltrasonicProximityGraph.reciprocityResidual`. */
export function reciprocityResidual(claim: ProximityClaim): number {
  return adist(claim.aToB, claim.bToA);
}

/** A claim is reciprocal at a tolerance when its directional ToFs agree up to
 *  `tol`. Mirrors `UltrasonicProximityGraph.reciprocalB`. */
export function reciprocal(tol: number, claim: ProximityClaim): boolean {
  return reciprocityResidual(claim) <= tol;
}

/* ------------------------------------------------------------------ */
/*  Gate 2: support / non-triviality                                 */
/* ------------------------------------------------------------------ */

/** A claim is trivial (no ranging information) when *both* directional ToFs are
 *  zero: nobody actually heard anybody. The source-only analog.
 *  Mirrors `UltrasonicProximityGraph.trivialClaimB`. */
export function trivialClaim(claim: ProximityClaim): boolean {
  return claim.aToB === 0 && claim.bToA === 0;
}

/** A claim carries support (is non-trivial) when it is not the all-zero claim:
 *  at least one real arrival was measured.
 *  Mirrors `UltrasonicProximityGraph.nonTrivialB`. */
export function nonTrivial(claim: ProximityClaim): boolean {
  return !trivialClaim(claim);
}

/* ------------------------------------------------------------------ */
/*  Gate 3: triangle consistency                                     */
/* ------------------------------------------------------------------ */

/**
 * A triangle of three devices given the pairwise one-way ToFs (proportional to
 * distance, speed normalized to 1). Mirrors `UltrasonicProximityGraph.Triangle`.
 */
export interface Triangle {
  readonly tofAB: number;
  readonly tofBC: number;
  readonly tofAC: number;
}

/**
 * The triangle is consistent at tolerance `tol` when the direct A->C ToF does
 * not exceed the around-the-corner path plus slack: `tofAC <= tofAB + tofBC + tol`.
 * A node faking closeness to A while measurably far from A's neighbour B inflates
 * `tofAC` and fails this. Accepts either the legs as separate arguments or a
 * `Triangle`. Mirrors `UltrasonicProximityGraph.triangleConsistentB`.
 *
 * @param ac  the direct A->C one-way ToF
 * @param ab  the A->B leg ToF
 * @param bc  the B->C leg ToF
 * @param tol the slack tolerance
 */
export function triangleConsistent(
  ac: number,
  ab: number,
  bc: number,
  tol: number,
): boolean {
  return ac <= ab + bc + tol;
}

/** Triangle-consistency over a `Triangle` record (the structured form). */
export function triangleConsistentTri(tol: number, tri: Triangle): boolean {
  return triangleConsistent(tri.tofAC, tri.tofAB, tri.tofBC, tol);
}

/* ------------------------------------------------------------------ */
/*  Runtime promotion check (epistemic outcome)                       */
/* ------------------------------------------------------------------ */

/**
 * Epistemic outcome of a proximity-edge observation. Identical vocabulary to the
 * sibling mesh carriers (`./cross-node-coincidence`,
 * `./electrical-network-frequency`), kept as a local literal union so this
 * certificate has no import coupling:
 *
 *  - `claim_admitted`    — the claim is reciprocal AND non-trivial, so the
 *                          proximity edge may be promoted to a graph edge.
 *  - `observation_only`  — reserved for a present-but-uncertifiable measurement:
 *                          a non-trivial but non-reciprocal claim is reported
 *                          (something was heard) yet refused as an edge. (Spoof
 *                          beacons land here: they carry support but break
 *                          reciprocity.)
 *  - `claim_withheld`    — the claim is trivial (zero ToF in both directions):
 *                          no ranging happened, so there is no edge to claim.
 */
export type UltrasonicProximityEpistemicOutcome =
  | 'claim_admitted'
  | 'observation_only'
  | 'claim_withheld';

/** The reason a proximity edge was held back from promotion. */
export type UltrasonicProximityLimitationClass =
  | 'no_ranging'
  | 'non_reciprocal_spoof'
  | 'none';

export interface UltrasonicProximityAdmission {
  readonly outcome: UltrasonicProximityEpistemicOutcome;
  readonly limitation: UltrasonicProximityLimitationClass;
  /** Whether the directional ToFs agreed within tolerance (reciprocity). */
  readonly reciprocal: boolean;
  /** Whether the claim carried any ranging support (non-trivial). */
  readonly nonTrivial: boolean;
  /** The measured reciprocity residual (the displayable asymmetry). */
  readonly reciprocityResidual: number;
  readonly summary: string;
}

export interface UltrasonicProximityAdmitOptions {
  /** Tolerance for the reciprocity residual. */
  readonly tol: number;
}

/**
 * The promotion check Monster Mesh runs before any UI surface displays a
 * proximity edge in the mesh graph. It folds the Lean contract into an epistemic
 * outcome:
 *
 *  - the support / non-triviality gate decides whether any ranging happened at
 *    all — a zero-ToF claim carries none and is withheld (the source-only
 *    rejection analog, mirroring `zero_tof_rejected`);
 *  - the reciprocity gate decides whether the measurement is consistent with two
 *    real co-present devices — a one-sided spoof beacon breaks reciprocity and is
 *    reported but never promoted (mirroring `spoof_not_reciprocal` /
 *    `spoof_rejected`);
 *  - only a reciprocal, non-trivial claim is admitted as a graph edge
 *    (mirroring `genuine_edge_admitted`).
 *
 * @param claim the reported proximity edge under examination
 * @param opts  the reciprocity tolerance
 */
export function admitProximityEdge(
  claim: ProximityClaim,
  opts: UltrasonicProximityAdmitOptions,
): UltrasonicProximityAdmission {
  const residual = reciprocityResidual(claim);
  const isReciprocal = residual <= opts.tol;
  const hasSupport = nonTrivial(claim);

  // Support gate: a zero-ToF claim carries no ranging information, so there is
  // no edge to claim even though it is (vacuously) reciprocal.
  if (!hasSupport) {
    return {
      outcome: 'claim_withheld',
      limitation: 'no_ranging',
      reciprocal: isReciprocal,
      nonTrivial: hasSupport,
      reciprocityResidual: residual,
      summary:
        'zero-ToF claim: nobody heard anybody (both directions 0) — no ranging, nothing to claim',
    };
  }

  // Reciprocity gate: a non-trivial but asymmetric claim is a spoof beacon — it
  // announced nearness without producing the reciprocal arrival.
  if (!isReciprocal) {
    return {
      outcome: 'observation_only',
      limitation: 'non_reciprocal_spoof',
      reciprocal: isReciprocal,
      nonTrivial: hasSupport,
      reciprocityResidual: residual,
      summary: `non-reciprocal claim: directional ToFs differ by ${residual} > tol ${opts.tol} — spoof beacon suspected, not an edge`,
    };
  }

  return {
    outcome: 'claim_admitted',
    limitation: 'none',
    reciprocal: isReciprocal,
    nonTrivial: hasSupport,
    reciprocityResidual: residual,
    summary: `reciprocal (residual ${residual} <= ${opts.tol}) and non-trivial — proximity edge certified`,
  };
}

/* ------------------------------------------------------------------ */
/*  Concrete fixtures (mirror the Lean fixtures)                      */
/* ------------------------------------------------------------------ */

/** A genuine reciprocal edge: chirp A->B took 12 ticks, B->A took 13, residual 1,
 *  within tolerance 2; both positive, so non-trivial. Admitted.
 *  Mirrors `UltrasonicProximityGraph.genuineEdge`. */
export const genuineEdge: ProximityClaim = {
  aToB: 12,
  bToA: 13,
  claimedDist: 12,
};

/** A spoof beacon: it announces nearness (forward ToF 12) but produced no
 *  matching reciprocal arrival — the back-direction ToF is 30, residual 18, far
 *  beyond tolerance 2. Rejected on reciprocity.
 *  Mirrors `UltrasonicProximityGraph.spoofEdge`. */
export const spoofEdge: ProximityClaim = {
  aToB: 12,
  bToA: 30,
  claimedDist: 12,
};

/** A zero-ToF claim: nobody actually heard anybody (both directions 0). It is
 *  reciprocal (residual 0) but carries no ranging support. Rejected on
 *  non-triviality. Mirrors `UltrasonicProximityGraph.zeroEdge`. */
export const zeroEdge: ProximityClaim = {
  aToB: 0,
  bToA: 0,
  claimedDist: 0,
};

/** A consistent triangle: A->B = 5, B->C = 4, A->C = 8. The direct hop 8 is
 *  within the around-the-corner path 5 + 4 + tol(1) = 10. Admitted.
 *  Mirrors `UltrasonicProximityGraph.consistentTriangle`. */
export const consistentTriangle: Triangle = { tofAB: 5, tofBC: 4, tofAC: 8 };

/** A violating triangle: the same legs (5, 4) but the device claims A->C = 20,
 *  impossibly larger than the path through B (5 + 4 + 1 = 10). The third-ear hard
 *  negative. Rejected. Mirrors `UltrasonicProximityGraph.violatingTriangle`. */
export const violatingTriangle: Triangle = { tofAB: 5, tofBC: 4, tofAC: 20 };

/**
 * Ultrasonic Proximity master witness. Bundles the boundary facts the Lean
 * theorem `ultrasonic_proximity_master` proves, each recomputed here from first
 * principles (never imported from Lean IR).
 *
 * Mirrors `UltrasonicProximityGraph.ultrasonic_proximity_master`.
 */
export const UltrasonicProximityMaster = {
  /** The genuine edge is admitted at tolerance 2. */
  genuineEdgeAdmitted:
    admitProximityEdge(genuineEdge, { tol: 2 }).outcome === 'claim_admitted',

  /** The spoof beacon fails reciprocity (residual 18 > 2). */
  spoofNotReciprocal: reciprocal(2, spoofEdge) === false,

  /** And so the spoof beacon is not admitted as an edge. */
  spoofRejected:
    admitProximityEdge(spoofEdge, { tol: 2 }).outcome !== 'claim_admitted',

  /** The zero-ToF claim is reciprocal (residual 0)... */
  zeroReciprocal: reciprocal(2, zeroEdge) === true,

  /** ...but trivial, so it is withheld (reciprocity alone is not enough). */
  zeroRejected:
    admitProximityEdge(zeroEdge, { tol: 2 }).outcome === 'claim_withheld',

  /** The consistent-triangle fixture passes the triangle gate at tolerance 1. */
  triangleConsistent:
    triangleConsistentTri(1, consistentTriangle) === true,

  /** The violating-triangle fixture fails the triangle gate at tolerance 1. */
  triangleViolationRejected:
    triangleConsistentTri(1, violatingTriangle) === false,

  /** The same announced forward ToF (12) yields opposite verdicts: forward range
   *  alone cannot be used as an edge key. */
  sameForwardTofOppositeVerdict:
    genuineEdge.aToB === spoofEdge.aToB &&
    admitProximityEdge(genuineEdge, { tol: 2 }).outcome === 'claim_admitted' &&
    admitProximityEdge(spoofEdge, { tol: 2 }).outcome !== 'claim_admitted',
} as const;
