// @a0n/gnosis/distributed-inference-host/src/visual-watermark-mesh-identity.ts
/**
 * Runtime mirror of `Gnosis.Mesh.VisualWatermarkMeshIdentity`.
 *
 * This carries the finite promotion contract only: a visual watermark claim is
 * admitted when residual correlation clears a threshold and the recovered token
 * is one of the mesh-issued tokens.
 */

export interface RecoveredVisualWatermark {
  readonly token: number;
  readonly correlation: number;
}

export type VisualWatermarkMeshEpistemicOutcome =
  | 'claim_admitted'
  | 'observation_only'
  | 'claim_withheld';

export type VisualWatermarkMeshLimitationClass =
  | 'weak_correlation'
  | 'forged_token'
  | 'none';

export interface VisualWatermarkMeshIdentityAdmission {
  readonly outcome: VisualWatermarkMeshEpistemicOutcome;
  readonly limitation: VisualWatermarkMeshLimitationClass;
  readonly correlationCleared: boolean;
  readonly tokenAdmitted: boolean;
  readonly summary: string;
}

export interface VisualWatermarkMeshIdentityAdmitOptions {
  readonly threshold: number;
}

/**
 * Handles the gnosis admitted Visual Watermark Token workflow.
 */
export function admittedVisualWatermarkToken(token: number): boolean {
  return token === 69 || token === 66 || token === 660;
}

/**
 * Handles the gnosis admit Visual Watermark Mesh Identity workflow.
 */
export function admitVisualWatermarkMeshIdentity(
  watermark: RecoveredVisualWatermark,
  options: VisualWatermarkMeshIdentityAdmitOptions,
): VisualWatermarkMeshIdentityAdmission {
  const correlationCleared = watermark.correlation >= options.threshold;
  const tokenAdmitted = admittedVisualWatermarkToken(watermark.token);
  if (correlationCleared && tokenAdmitted) {
    return {
      outcome: 'claim_admitted',
      limitation: 'none',
      correlationCleared,
      tokenAdmitted,
      summary: 'visual watermark provenance admitted',
    };
  }
  const limitation: VisualWatermarkMeshLimitationClass = correlationCleared
    ? 'forged_token'
    : 'weak_correlation';
  return {
    outcome: 'claim_withheld',
    limitation,
    correlationCleared,
    tokenAdmitted,
    summary:
      limitation === 'forged_token'
        ? 'visual watermark token was not issued by the mesh'
        : 'visual watermark correlation is below the admission threshold',
  };
}

/**
 * Handles the gnosis visual Watermark Mesh Identity Master Runtime workflow.
 */
export function visualWatermarkMeshIdentityMasterRuntime(): {
  readonly genuineAdmitted: boolean;
  readonly forgedRejected: boolean;
  readonly weakRejected: boolean;
} {
  const threshold = 0.5;
  const genuine = admitVisualWatermarkMeshIdentity(
    { token: 69, correlation: 0.9 },
    { threshold },
  );
  const forged = admitVisualWatermarkMeshIdentity(
    { token: 7, correlation: 0.9 },
    { threshold },
  );
  const weak = admitVisualWatermarkMeshIdentity(
    { token: 69, correlation: 0.1 },
    { threshold },
  );
  return {
    genuineAdmitted: genuine.outcome === 'claim_admitted',
    forgedRejected: forged.outcome === 'claim_withheld',
    weakRejected: weak.outcome === 'claim_withheld',
  };
}
