// @a0n/gnosis/distributed-inference-host/src/vital-sign-micro-doppler.ts
/**
 * # Vital-Sign Micro-Doppler — source-subtracted breathing residue as a vital sign
 *
 * Mirrors `Gnosis.Mesh.VitalSignMicroDoppler` from `open-source/gnosis-math`,
 * the Rustic Church carrier for the BGM2Pose cousin: a body's breathing and
 * heartbeat micro-motion faintly modulates an ambient acoustic carrier. After
 * the steady room-tone / tracked-source baseline is subtracted, what is left is
 * a slow, periodic *residue envelope* — the vital sign. A static body (no chest
 * motion) modulates nothing, so its subtracted envelope is flat and carries no
 * reading.
 *
 * ## What is (and is not) carried
 *
 * This module is the runtime promotion check for a vital-sign observation. It
 * carries ONLY the finite contract the Lean file proves, three gates:
 *
 *  - the saturating residue metric (`adist`);
 *  - the support gate (the source-only rejection analog): a below-floor total
 *    residue magnitude is a body that modulated nothing — no signal to read
 *    (`totalMagnitude` / support floor);
 *  - the liveness / variation gate: a constant residue has zero total variation,
 *    a static body, no breathing rhythm (`totalVariation`);
 *  - the plausible-rate band gate: the residue peak count (the breathing rhythm)
 *    must lie in a physiological band `[minPeaks, maxPeaks]` — too few is no
 *    rhythm, too many is noise, not breathing (`peakCount`).
 *
 * Real STFT micro-Doppler extraction, baseline tracking, true respiratory /
 * cardiac rate estimation, motion-artifact rejection, and SNR robustness stay
 * runtime/empirical obligations OUTSIDE this certificate — they are not theorems
 * and are not asserted here.
 *
 * The promotion check folds the contract into the same epistemic-outcome
 * vocabulary the sibling mesh carriers already speak
 * (`claim_admitted` / `observation_only` / `claim_withheld`), so a vital-sign
 * observation cannot be displayed as a reading until it has cleared the
 * Lean-side support/liveness/rate-band gate. The hard negative is the *same body
 * in a different vital state*: a static (constant) envelope and a breathing
 * (periodic) envelope come from one body, and only the modulation distinguishes
 * the reading — source identity alone cannot serve as a vital-sign key.
 *
 * ## Lean source-of-truth
 *
 * @see open-source/gnosis-math/Gnosis/Mesh/VitalSignMicroDoppler.lean
 * @see open-source/gnosis-math/Gnosis/Mesh/ElectricalNetworkFrequency.lean (variation sibling)
 * @see open-source/gnosis-math/Gnosis/Mesh/DopplerSpeed.lean (motion sibling)
 */

/* ------------------------------------------------------------------ */
/*  Residue distance and envelopes                                    */
/* ------------------------------------------------------------------ */

/**
 * A residue envelope is a list of per-frame residue magnitudes above the tracked
 * baseline (already source-subtracted). Magnitudes are non-negative integers
 * (saturating `Nat` arithmetic), matching the Lean `Envelope`.
 */
export type Envelope = readonly number[];

/** Absolute difference on `Nat` via saturating subtraction (`|a - b|`). The
 *  per-frame residue distance metric, shared with the mesh corridor carriers.
 *  Mirrors `VitalSignMicroDoppler.adist`. */
export function adist(a: number, b: number): number {
  return Math.max(0, a - b) + Math.max(0, b - a);
}

/** Total residue magnitude across the window — the quantity the support gate
 *  sums. Mirrors `VitalSignMicroDoppler.totalMagnitude`. */
export function totalMagnitude(env: Envelope): number {
  let sum = 0;
  for (const x of env) {
    sum += x;
  }
  return sum;
}

/**
 * Total variation of an envelope: the summed residue distance between successive
 * frames. A flat envelope has zero total variation; a modulating one has
 * positive total variation. Mirrors `VitalSignMicroDoppler.totalVariation`.
 */
export function totalVariation(env: Envelope): number {
  let tv = 0;
  for (let i = 1; i < env.length; i++) {
    tv += adist(env[i - 1], env[i]);
  }
  return tv;
}

/**
 * Count residue peaks (local maxima): a frame strictly greater than both of its
 * neighbors. The breathing rate is this rhythm count. Endpoints are never peaks
 * (they have only one neighbor), matching the Lean triple recursion.
 * Mirrors `VitalSignMicroDoppler.peakCount`.
 */
export function peakCount(env: Envelope): number {
  let count = 0;
  for (let i = 1; i + 1 < env.length; i++) {
    if (env[i - 1] < env[i] && env[i + 1] < env[i]) {
      count += 1;
    }
  }
  return count;
}

/* ------------------------------------------------------------------ */
/*  Runtime promotion check (epistemic outcome)                       */
/* ------------------------------------------------------------------ */

/**
 * Epistemic outcome of a vital-sign observation. Identical vocabulary to the
 * sibling mesh carriers (`./electrical-network-frequency`,
 * `./cross-node-coincidence`), kept as a local literal union so this certificate
 * has no import coupling:
 *
 *  - `claim_admitted`    — a supported, varying envelope whose rhythm lands in
 *                          the plausible-rate band, so the vital sign may be
 *                          promoted to a reading.
 *  - `observation_only`  — a supported, varying envelope (motion is present) but
 *                          the peak count is out of the physiological band: a
 *                          motion is reported, never promoted to a clean vital
 *                          rhythm.
 *  - `claim_withheld`    — the envelope is below the support floor (a body that
 *                          modulated nothing) OR it is supported but static
 *                          (zero variation, no breathing rhythm), so there is no
 *                          reading to claim.
 */
export type VitalSignEpistemicOutcome =
  | 'claim_admitted'
  | 'observation_only'
  | 'claim_withheld';

/** The reason a vital-sign reading was held back from promotion. */
export type VitalSignLimitationClass =
  | 'no_support'
  | 'static_body'
  | 'rhythm_out_of_band'
  | 'none';

export interface VitalSignAdmission {
  readonly outcome: VitalSignEpistemicOutcome;
  readonly limitation: VitalSignLimitationClass;
  /** Whether the residue cleared the support floor. */
  readonly supported: boolean;
  /** Whether the residue actually varies (passed the liveness gate). */
  readonly varied: boolean;
  /** The measured total residue magnitude (the displayable support level). */
  readonly totalMagnitude: number;
  /** The measured total variation (the displayable modulation magnitude). */
  readonly totalVariation: number;
  /** The measured peak count (the displayable breathing rhythm). */
  readonly peakCount: number;
  readonly summary: string;
}

export interface VitalSignAdmitOptions {
  /** Support floor: the total residue magnitude must exceed this. */
  readonly floor: number;
  /** Lower bound of the physiological rate band (inclusive). */
  readonly minPeaks: number;
  /** Upper bound of the physiological rate band (inclusive). */
  readonly maxPeaks: number;
}

/**
 * The promotion check Monster Mesh runs before any UI surface displays a
 * vital-sign observation as a reading. It folds the Lean contract into an
 * epistemic outcome:
 *
 *  - the support gate decides whether there is any signal at all — a below-floor
 *    envelope carries no vital sign and is withheld (the source-only rejection
 *    analog, mirroring `zero_envelope_rejected`);
 *  - the liveness gate decides whether the body modulated anything — a static
 *    (constant) envelope has zero variation, no breathing rhythm, and is
 *    withheld (mirroring `static_body_rejected`);
 *  - the rate-band gate decides whether a present modulation is a *clean* vital
 *    rhythm — motion whose peak count falls outside `[minPeaks, maxPeaks]` is
 *    reported as observation-only rather than a promotable reading (mirroring
 *    `noisy_out_of_band_rejected`).
 *
 * @param env  the source-subtracted residue envelope under examination
 * @param opts the support floor and physiological rate band
 */
export function admitVitalSign(
  env: Envelope,
  opts: VitalSignAdmitOptions,
): VitalSignAdmission {
  const mag = totalMagnitude(env);
  const tv = totalVariation(env);
  const peaks = peakCount(env);
  const supported = mag > opts.floor;
  const varied = tv > 0;
  const inBand = peaks >= opts.minPeaks && peaks <= opts.maxPeaks;

  // Support gate: a below-floor residue is a body that modulated nothing above
  // the noise floor — there is no signal to read.
  if (!supported) {
    return {
      outcome: 'claim_withheld',
      limitation: 'no_support',
      supported,
      varied,
      totalMagnitude: mag,
      totalVariation: tv,
      peakCount: peaks,
      summary:
        'below-floor residue: no support above the noise floor — nothing to read',
    };
  }

  // Liveness gate: a supported but static residue is a body with no chest
  // motion — a perfectly steady subtracted baseline, no breathing rhythm.
  if (!varied) {
    return {
      outcome: 'claim_withheld',
      limitation: 'static_body',
      supported,
      varied,
      totalMagnitude: mag,
      totalVariation: tv,
      peakCount: peaks,
      summary:
        'static body: supported residue with zero variation carries no breathing rhythm — nothing to read',
    };
  }

  // Rate-band gate: a supported, varying residue whose rhythm is out of the
  // physiological band is motion present but not a clean vital rhythm.
  if (!inBand) {
    return {
      outcome: 'observation_only',
      limitation: 'rhythm_out_of_band',
      supported,
      varied,
      totalMagnitude: mag,
      totalVariation: tv,
      peakCount: peaks,
      summary: `motion present (${peaks} peaks) but outside the rate band [${opts.minPeaks},${opts.maxPeaks}] — not a clean vital rhythm`,
    };
  }

  return {
    outcome: 'claim_admitted',
    limitation: 'none',
    supported,
    varied,
    totalMagnitude: mag,
    totalVariation: tv,
    peakCount: peaks,
    summary: `supported, varying residue with ${peaks} peaks in band [${opts.minPeaks},${opts.maxPeaks}] — vital sign read`,
  };
}

/* ------------------------------------------------------------------ */
/*  Finite certificate                                                */
/* ------------------------------------------------------------------ */

/**
 * A finite certificate that Monster Mesh has all required vital-sign gates.
 * Mirrors `VitalSignMicroDoppler.VitalSignCertificate`.
 */
export interface VitalSignCertificate {
  readonly residueModeled: boolean;
  readonly supportGated: boolean;
  readonly variationRequired: boolean;
  readonly rateBandGated: boolean;
  readonly staticRejected: boolean;
}

/** The vital-sign micro-Doppler carrier for Monster Mesh promotion checks.
 *  Mirrors `VitalSignMicroDoppler.vitalSignRusticCertificate`. */
export const vitalSignRusticCertificate: VitalSignCertificate = {
  residueModeled: true,
  supportGated: true,
  variationRequired: true,
  rateBandGated: true,
  staticRejected: true,
};

/* ------------------------------------------------------------------ */
/*  Concrete fixtures (mirror the Lean fixtures)                      */
/* ------------------------------------------------------------------ */

/** A breathing body: a clear, slow periodic residue with three peaks in band.
 *  totalMagnitude 24, totalVariation 20, peakCount 3. Mirrors `breathingEnv`. */
export const breathingEnv: Envelope = [2, 5, 2, 6, 2, 5, 2];

/** A static body, same subject: a flat residue. No modulation, no reading.
 *  totalMagnitude 28, totalVariation 0, peakCount 0. Mirrors `staticEnv`. */
export const staticEnv: Envelope = [4, 4, 4, 4, 4, 4, 4];

/** A noisy residue: chatter, not breathing. Four peaks — above the rate band.
 *  totalMagnitude 41, totalVariation 64, peakCount 4. Mirrors `noisyEnv`. */
export const noisyEnv: Envelope = [1, 9, 1, 9, 1, 9, 1, 9, 1];

/** A below-floor residue: the body modulated nothing above the noise floor.
 *  totalMagnitude 0. Mirrors `zeroEnv`. */
export const zeroEnv: Envelope = [0, 0, 0, 0];

/**
 * Vital-Sign Micro-Doppler master witness. Bundles the boundary facts the Lean
 * theorem `vital_sign_master` proves, each recomputed here from first principles
 * (never imported from Lean IR). Gate parameters: floor 3, band [1,3].
 *
 * Mirrors `VitalSignMicroDoppler.vital_sign_master`.
 */
export const VitalSignMaster = {
  /** A breathing body is admitted: supported, varying, in-band. */
  breathingAdmitted:
    admitVitalSign(breathingEnv, { floor: 3, minPeaks: 1, maxPeaks: 3 })
      .outcome === 'claim_admitted',

  /** A static body is rejected by the liveness gate. */
  staticRejected:
    admitVitalSign(staticEnv, { floor: 3, minPeaks: 1, maxPeaks: 3 })
      .outcome !== 'claim_admitted',

  /** A noisy residue is not admitted: its rhythm is out of the rate band. */
  noisyNotAdmitted:
    admitVitalSign(noisyEnv, { floor: 3, minPeaks: 1, maxPeaks: 3 })
      .outcome !== 'claim_admitted',

  /** A below-floor residue is rejected: there is no support to read. */
  zeroRejected:
    admitVitalSign(zeroEnv, { floor: 3, minPeaks: 1, maxPeaks: 3 })
      .outcome === 'claim_withheld',

  /** The breathing envelope clears the support floor (magnitude 24 > 3). */
  breathingSupported: totalMagnitude(breathingEnv) > 3,

  /** The breathing envelope varies (total variation 20 > 0). */
  breathingVaried: totalVariation(breathingEnv) > 0,

  /** The breathing envelope's rhythm (3 peaks) lands in band [1,3]. */
  breathingInBand:
    peakCount(breathingEnv) >= 1 && peakCount(breathingEnv) <= 3,

  /** The hard negative: same body, breathing admitted, static rejected. */
  sameBodyDifferentVitalState:
    admitVitalSign(breathingEnv, { floor: 3, minPeaks: 1, maxPeaks: 3 })
      .outcome === 'claim_admitted' &&
    admitVitalSign(staticEnv, { floor: 3, minPeaks: 1, maxPeaks: 3 }).outcome !==
      'claim_admitted',

  /** The certificate records the rate-band gate. */
  certRateBandGated: vitalSignRusticCertificate.rateBandGated,

  /** The certificate records static rejection. */
  certStaticRejected: vitalSignRusticCertificate.staticRejected,
} as const;
