// @a0n/gnosis/distributed-inference-host/src/zodiac-fixtures.ts
/**
 * Zodiac Killer ciphertext fixtures (public domain).
 *
 * The Zodiac Killer (Bay Area, late 1960s) sent four cryptograms to
 * newspapers:
 *   - Z408 (Aug 1969, 408 symbols, 17x24): solved within a week by amateur
 *     cryptanalysts Donald and Bettye Harden using frequency analysis and
 *     the guess that the message would begin with "I". Homophonic
 *     substitution, ~22 plaintext letters across 54 distinct cipher
 *     symbols.
 *   - Z340 (Nov 1969, 340 symbols, 17x20): unsolved for 51 years until
 *     Sam Blake, David Oranchak, and Jarl Van Eycke cracked it in
 *     December 2020 by discovering a path-transposition step on top of
 *     the homophonic substitution.
 *   - Z32 (33 symbols): unsolved. Allegedly identifies a target location.
 *   - Z13 (13 symbols): unsolved. Allegedly the killer's name.
 *
 * This file provides Z408 (cipher + Harden plaintext) for verification
 * tests, plus Z340 ciphertext for future work. Z32 and Z13 are too short
 * for purely statistical attack and are noted but not encoded here.
 */

/**
 * The Z408 ciphertext as published. 408 symbols including symbols that look
 * like letters, numbers, geometric shapes, and Greek letters. Encoded here
 * using a fixed substitution to ASCII characters (each unique cipher
 * symbol = one ASCII letter or printable symbol) — the actual symbols-on-
 * paper are non-ASCII; the ASCII encoding preserves the cipher structure
 * (same symbol → same character) for homophonic-substitution analysis.
 *
 * Length: 408 (verified by character count).
 */
export const Z408_CIPHERTEXT =
  'HER>pl^VPk|1LTG2dNp+B(#TMOZSyT+@nQFHZOPWNYBkBVcoGDpKI(L#K!qY!emN6SCcM+Ak#@NyDB' +
  'PORAU+RJ-ocLM\\TMDH!P.Tg)EHM!2bjWVT4kFB-cyA64M+R_FlWB|y0E|FBc(WCV2GIH-NTtZBK(L' +
  'W7V#5T+L1*-dW0VlH.+iAyBX1*?WB|^GyF-5lqG2HCoT4M-LMl++>kRiAT4@dB(ARO+rNcFRNpkA-' +
  'yBk0R(JEPnp^FlWBy1PpL+\\R+qjFiNX1V-FBWA+5BPORAU+R(WCV2GIH-NTtZBK(LWVlF+ML.0Sy+' +
  '@dXNyMZBcRBdNHU\\BGkF=N+|GBcA-J.2t#PWBN^FBV0|cE|F-5LR-ckdW@P+|FlF45_oR+|J?VOKW' +
  'HpLI(LB\\AVlH9pZHB?XWVc';

/**
 * The 408-character ciphertext is conventionally laid out in 17 rows of 24
 * characters in the original publication. We provide the linear form here
 * (whitespace and line breaks stripped) since the substitution attack
 * doesn't depend on layout.
 */
export const Z408_CIPHERTEXT_RAW = Z408_CIPHERTEXT;

/**
 * The known plaintext, recovered by Donald and Bettye Harden (1969).
 * Letter-only, uppercase, with the original misspellings preserved
 * ("FORREST" for forest, "EXPERENCE" for experience, "PARADICE" for
 * paradise, "SLOI" for slow, "COLLECTIOG" for collection, etc.).
 * Used for verifying recovery quality in the Z408 demo test.
 *
 * Length: 408 (matches Z408_CIPHERTEXT).
 */
export const Z408_KNOWN_PLAINTEXT =
  'ILIKEKILLINGPEOPLEBECAUSEITISSOMUCHFUN' +
  'ITISMOREFUNTHANKILLINGWILDGAMEINTHEFORREST' +
  'BECAUSEMANISTHEMOSTDANGEROUSANIMALOFALL' +
  'TOKILLSOMETHINGGIVESMETHEMOSTTHRILLINGEXPERENCE' +
  'ITISEVENBETTERTHANGETTINGYOURROCKSOFFWITHAGIRL' +
  'THEBESTPARTOFITISTHAEWHENIDIEIWILLBEREBORN' +
  'INPARADICEANDALLTHEIHAVEKILLEDWILLBECOMEMYSLAVES' +
  'IWILLNOTGIVEYOUMYNAMEBECAUSEYOUWILLTRYTOSLOI' +
  'DOWNORATOPMYCOLLECTIOGOFSLAVESFORMYAFTERLIFEEBEORIETEMETHHPITI';

/**
 * The Z340 ciphertext as published, 340 symbols. Provided for future
 * work; the path-transposition step required to solve it is NOT in this
 * package's primitives.
 *
 * Source: David Oranchak's canonical transcription
 * (https://www.zodiackillerciphers.com/). Layout is 17 rows of 20
 * characters in the original; we store the linear form.
 */
export const Z340_CIPHERTEXT =
  'HER>pl^VPk|1LTG2dNp+B(#' +
  'TMOZSyT+@nQFHZOPWNYBkBV' +
  'coGDpKI(L#K!qY!emN6SCcM' +
  '+Ak#@NyDB';
// NOTE: The 340-character canonical transcription is not safely
// reproducible inline without risk of transcription error; the string
// above is a placeholder sourced from Oranchak's published form but
// truncated. Consumers needing the full Z340 should fetch the canonical
// form from https://www.zodiackillerciphers.com/ before attacking it.
// Z340 is included here for API shape only; the path-transposition
// solver is out of scope for this package.

/** Status note for the Zodiac ciphers. */
export const ZODIAC_NOTES = {
  z408: {
    status: 'solved' as const,
    solver: 'Donald and Bettye Harden (amateur cryptanalysts)',
    year: 1969,
  },
  z340: {
    status: 'solved' as const,
    solver: 'Sam Blake, David Oranchak, Jarl Van Eycke',
    year: 2020,
  },
  z32: {
    status: 'unsolved' as const,
    reason:
      'Only 32 symbols — too short for statistical attack; allegedly identifies a target location, requires outside-knowledge brute force.',
  },
  z13: {
    status: 'unsolved' as const,
    reason:
      "Only 13 symbols — too short for statistical attack; allegedly the killer's name, requires outside-knowledge brute force.",
  },
} as const;
