// @a0n/gnosis/forkjoin-app/client/hooks/useRichPresenceTelemetry.ts
import { useEffect, useRef } from '@a0n/raect';
import type {
  PresenceEmotion,
  PresenceInputState,
  PresenceScroll,
  PresenceSelection,
  PresenceTyping,
  PresenceViewport,
} from './usePresence';

interface RichPresenceTelemetryOptions {
  enabled?: boolean;
  sendFocusNode: (focusNode: string) => void;
  sendSelection: (selection: PresenceSelection) => void;
  sendTyping: (typing: PresenceTyping) => void;
  sendScroll: (scroll: PresenceScroll) => void;
  sendViewport: (viewport: PresenceViewport) => void;
  sendInputState: (inputState: PresenceInputState) => void;
  sendEmotionState: (emotion: PresenceEmotion) => void;
}

/**
 * Provides the use Rich Presence Telemetry hook.
 */
export function useRichPresenceTelemetry({
  enabled = true,
  sendFocusNode,
  sendSelection,
  sendTyping,
  sendScroll,
  sendViewport,
  sendInputState,
  sendEmotionState,
}: RichPresenceTelemetryOptions) {
  const typingStopTimerRef = useRef<ReturnType<
    typeof window.setTimeout
  > | null>(null);
  const scrollFrameRef = useRef<ReturnType<
    typeof window.requestAnimationFrame
  > | null>(null);

  useEffect(() => {
    if (!enabled) return;

    sendViewport({ width: window.innerWidth, height: window.innerHeight });
    sendEmotionState({
      primary: 'curious',
      source: 'inferred',
      confidence: 0.35,
      updatedAt: new Date().toISOString(),
    });

    const handleResize = () => {
      sendViewport({ width: window.innerWidth, height: window.innerHeight });
    };

    const handleScroll = () => {
      if (scrollFrameRef.current != null) return;
      scrollFrameRef.current = window.requestAnimationFrame(() => {
        scrollFrameRef.current = null;
        const root = document.documentElement;
        const maxScroll = Math.max(1, root.scrollHeight - window.innerHeight);
        const y = window.scrollY;
        sendScroll({
          depth: Math.max(0, Math.min(1, y / maxScroll)),
          y,
          viewportHeight: window.innerHeight,
          documentHeight: root.scrollHeight,
          path: window.location.pathname,
        });
      });
    };

    const handleFocusIn = (event: FocusEvent) => {
      const target = event.target;
      if (!(target instanceof HTMLElement)) return;
      sendFocusNode(resolveNodePath(target));

      const field = resolveFieldName(target);
      if (field) {
        sendInputState({
          field,
          hasFocus: true,
          valueLength: getElementValueLength(target),
          selectionStart: getSelectionStart(target),
          selectionEnd: getSelectionEnd(target),
          inputMode: getInputMode(target),
        });
      }
    };

    const handleFocusOut = (event: FocusEvent) => {
      const target = event.target;
      if (!(target instanceof HTMLElement)) return;

      const field = resolveFieldName(target);
      if (field) {
        sendInputState({
          field,
          hasFocus: false,
          valueLength: getElementValueLength(target),
          selectionStart: getSelectionStart(target),
          selectionEnd: getSelectionEnd(target),
          inputMode: getInputMode(target),
        });
      }
    };

    const handleSelectionChange = () => {
      const active = document.activeElement;
      if (
        !(active instanceof HTMLInputElement) &&
        !(active instanceof HTMLTextAreaElement)
      ) {
        return;
      }

      if (
        typeof active.selectionStart !== 'number' ||
        typeof active.selectionEnd !== 'number'
      ) {
        return;
      }

      sendSelection({
        start: active.selectionStart,
        end: active.selectionEnd,
        direction: active.selectionDirection ?? undefined,
        path: resolveNodePath(active),
      });
    };

    const scheduleTypingStop = (field: string) => {
      if (typingStopTimerRef.current != null) {
        window.clearTimeout(typingStopTimerRef.current);
      }
      typingStopTimerRef.current = setTimeout(() => {
        sendTyping({
          isTyping: false,
          field,
          stoppedAt: new Date().toISOString(),
        });
      }, 1200);
    };

    const handleInput = (event: Event) => {
      const target = event.target;
      if (!(target instanceof HTMLElement)) return;
      const field = resolveFieldName(target);
      if (!field) return;

      const inputState: PresenceInputState = {
        field,
        hasFocus: true,
        valueLength: getElementValueLength(target),
        selectionStart: getSelectionStart(target),
        selectionEnd: getSelectionEnd(target),
        inputMode: getInputMode(target),
      };
      sendInputState(inputState);
      sendTyping({
        isTyping: true,
        field,
        startedAt: new Date().toISOString(),
      });
      scheduleTypingStop(field);
    };

    const handleCompositionStart = (event: CompositionEvent) => {
      const target = event.target;
      if (!(target instanceof HTMLElement)) return;
      const field = resolveFieldName(target);
      if (!field) return;

      sendTyping({
        isTyping: true,
        field,
        isComposing: true,
        startedAt: new Date().toISOString(),
      });
      sendInputState({
        field,
        hasFocus: true,
        isComposing: true,
        valueLength: getElementValueLength(target),
        selectionStart: getSelectionStart(target),
        selectionEnd: getSelectionEnd(target),
        inputMode: getInputMode(target),
      });
    };

    const handleCompositionEnd = (event: CompositionEvent) => {
      const target = event.target;
      if (!(target instanceof HTMLElement)) return;
      const field = resolveFieldName(target);
      if (!field) return;

      sendTyping({
        isTyping: true,
        field,
        isComposing: false,
      });
      sendInputState({
        field,
        hasFocus: true,
        isComposing: false,
        valueLength: getElementValueLength(target),
        selectionStart: getSelectionStart(target),
        selectionEnd: getSelectionEnd(target),
        inputMode: getInputMode(target),
      });
      scheduleTypingStop(field);
    };

    const handleVisibilityChange = () => {
      sendEmotionState({
        primary: document.hidden ? 'away' : 'engaged',
        source: 'inferred',
        confidence: 0.4,
        updatedAt: new Date().toISOString(),
      });
    };

    window.addEventListener('resize', handleResize, { passive: true });
    window.addEventListener('scroll', handleScroll, { passive: true });
    document.addEventListener('focusin', handleFocusIn);
    document.addEventListener('focusout', handleFocusOut);
    document.addEventListener('selectionchange', handleSelectionChange);
    document.addEventListener('input', handleInput);
    document.addEventListener('compositionstart', handleCompositionStart);
    document.addEventListener('compositionend', handleCompositionEnd);
    document.addEventListener('visibilitychange', handleVisibilityChange);

    handleScroll();

    return () => {
      window.removeEventListener('resize', handleResize);
      window.removeEventListener('scroll', handleScroll);
      document.removeEventListener('focusin', handleFocusIn);
      document.removeEventListener('focusout', handleFocusOut);
      document.removeEventListener('selectionchange', handleSelectionChange);
      document.removeEventListener('input', handleInput);
      document.removeEventListener('compositionstart', handleCompositionStart);
      document.removeEventListener('compositionend', handleCompositionEnd);
      document.removeEventListener('visibilitychange', handleVisibilityChange);
      if (typingStopTimerRef.current != null) {
        window.clearTimeout(typingStopTimerRef.current);
      }
      if (scrollFrameRef.current != null) {
        window.cancelAnimationFrame(scrollFrameRef.current);
      }
    };
  }, [
    enabled,
    sendEmotionState,
    sendFocusNode,
    sendInputState,
    sendScroll,
    sendSelection,
    sendTyping,
    sendViewport,
  ]);
}

function resolveFieldName(element: HTMLElement): string | null {
  if (element.dataset.presenceField) return element.dataset.presenceField;
  if ('name' in element && typeof element.name === 'string' && element.name) {
    return element.name;
  }
  if (element.id) return element.id;
  if (element.getAttribute('aria-label')) {
    return element.getAttribute('aria-label');
  }
  return element.tagName.toLowerCase();
}

function resolveNodePath(element: HTMLElement): string {
  const path: string[] = [];
  let node: HTMLElement | null = element;

  while (node && path.length < 6) {
    if (node.id) {
      path.unshift(`#${node.id}`);
      break;
    }

    const tag = node.tagName.toLowerCase();
    const className = node.className.toString().split(/\s+/).filter(Boolean)[0];
    const descriptor = className ? `${tag}.${className}` : tag;
    path.unshift(descriptor);
    node = node.parentElement;
  }

  return path.join(' > ');
}

function getElementValueLength(element: HTMLElement): number | undefined {
  if (
    element instanceof HTMLInputElement ||
    element instanceof HTMLTextAreaElement
  ) {
    return element.value.length;
  }
  if (element.isContentEditable) {
    return element.textContent?.length ?? 0;
  }
  return undefined;
}

function getSelectionStart(element: HTMLElement): number | undefined {
  if (
    element instanceof HTMLInputElement ||
    element instanceof HTMLTextAreaElement
  ) {
    return element.selectionStart ?? undefined;
  }
  return undefined;
}

function getSelectionEnd(element: HTMLElement): number | undefined {
  if (
    element instanceof HTMLInputElement ||
    element instanceof HTMLTextAreaElement
  ) {
    return element.selectionEnd ?? undefined;
  }
  return undefined;
}

function getInputMode(element: HTMLElement): string | undefined {
  if (
    element instanceof HTMLInputElement ||
    element instanceof HTMLTextAreaElement
  ) {
    return element.inputMode || undefined;
  }
  return undefined;
}
