// @a0n/.ignored_gnosis/distributed-inference-host/src/acoustic-replay-liveness.ts
/**
 * # Acoustic Replay Liveness — one room vs two convolved rooms
 *
 * Mirrors `Gnosis.Mesh.AcousticReplayLiveness` from `open-source/gnosis-math`,
 * the Rustic Church carrier for replay / liveness detection in the Monster Mesh
 * acoustic surface. The intuition is physical:
 *
 *  - A **live** sound arriving at a mic carries exactly ONE room's acoustic
 *    signature — the reverberation tail of the room it was emitted in (its
 *    impulse-response tail).
 *  - A **replayed** recording (a deepfake injected via a loudspeaker) carries
 *    TWO rooms convolved: the room it was originally recorded in AND the room it
 *    is now being played back in. Convolving two impulse responses *lengthens*
 *    the reflection tail, so a replay can never produce a shorter tail than
 *    either of its constituent rooms.
 *
 * The Skeptic admits *liveness* only when the observed reverberation is
 * consistent with a single plausible room — i.e. its reflection span fits inside
 * the bound a single room can have.
 *
 * ## What is (and is not) carried
 *
 * This module is the runtime promotion check for an acoustic liveness
 * observation. It carries ONLY the finite contract the Lean file proves:
 *
 *  - model the impulse response by its reflection span (taps) as a non-negative
 *    integer;
 *  - model two-room convolution by `convSpan a b = a + b - 1` (spans add, minus
 *    the one shared onset tap);
 *  - composition never shortens the tail (`convSpan_ge_left` /
 *    `convSpan_ge_right`);
 *  - a single room within bound is admitted as live (`live_room_admitted`);
 *  - a two-room composition exceeding the single-room bound is rejected as a
 *    replay (`replay_rejected_fixture`);
 *  - the SAME source through one room (live) vs two rooms (replayed) lands on
 *    opposite sides of the gate — the structural hard negative
 *    (`same_source_different_liveness`).
 *
 * Real STFT, real IR estimation, dereverberation, multipath, and loudspeaker
 * coloration stay runtime/empirical obligations OUTSIDE this certificate — they
 * are not theorems and are not asserted here.
 *
 * The promotion check folds the contract into the same epistemic-outcome
 * vocabulary the sibling mesh carriers already speak
 * (`claim_admitted` / `claim_withheld`), so an acoustic observation cannot be
 * displayed as a live-source claim until it has cleared the Lean-side
 * single-room bound.
 *
 * ## Lean source-of-truth
 *
 * @see open-source/gnosis-math/Gnosis/Mesh/AcousticReplayLiveness.lean
 * @see open-source/gnosis-math/Gnosis/Mesh/CrossNodeCoincidence.lean (mesh sibling)
 */

/* ------------------------------------------------------------------ */
/*  Rooms, reflection spans, and convolution                          */
/* ------------------------------------------------------------------ */

/**
 * A room's acoustic signature, modeled by the length of its reverberation tail
 * in taps. Any real room has `reflectionSpan >= 1` (at least the onset tap).
 * Mirrors `AcousticReplayLiveness.RoomEcho`.
 */
export interface RoomEcho {
  readonly reflectionSpan: number;
}

/**
 * Composing two rooms (recording room ⊛ playback room): the spans add, minus the
 * single shared onset tap. This is the discrete-convolution support length.
 * Mirrors `AcousticReplayLiveness.convSpan`.
 *
 * Saturated at zero to match `Nat` truncated subtraction (degenerate empty
 * spans cannot produce a negative tail).
 */
export function convSpan(a: number, b: number): number {
  return Math.max(0, a + b - 1);
}

/* ------------------------------------------------------------------ */
/*  The liveness gate                                                 */
/* ------------------------------------------------------------------ */

/**
 * The maximum reflection span a single plausible room can have. Above this the
 * observed tail can only be explained by two convolved rooms — a replay.
 * Mirrors `AcousticReplayLiveness.singleRoomBound`.
 */
export const singleRoomBound = 8;

/** Executable liveness admission: the observed span fits inside the bound, so it
 *  is consistent with a single room. Mirrors `admitsLiveB` / `AdmitsLive`. */
export function admitsLive(bound: number, span: number): boolean {
  return span <= bound;
}

/* ------------------------------------------------------------------ */
/*  Runtime promotion check (epistemic outcome)                       */
/* ------------------------------------------------------------------ */

/**
 * Epistemic outcome of an acoustic liveness observation. Identical vocabulary to
 * the sibling mesh carriers (`./cross-node-coincidence`,
 * `./active-acoustic-pose`), kept as a local literal union so this certificate
 * has no import coupling:
 *
 *  - `claim_admitted`    — the observed reflection span fits inside the
 *                          single-room bound, so the sound is consistent with
 *                          one plausible room and may be promoted to a
 *                          live-source claim.
 *  - `claim_withheld`    — the observed span exceeds the single-room bound, so
 *                          it can only be explained by two convolved rooms (a
 *                          replay); the liveness claim is withheld.
 */
export type LivenessEpistemicOutcome = 'claim_admitted' | 'claim_withheld';

/** The reason a liveness claim was held back from promotion. */
export type LivenessLimitationClass = 'replay_composition_suspected' | 'none';

export interface LivenessAdmission {
  readonly outcome: LivenessEpistemicOutcome;
  readonly limitation: LivenessLimitationClass;
  /** The observed reflection span being tested against the bound. */
  readonly reflectionSpan: number;
  /** The single-room bound applied. */
  readonly singleRoomBound: number;
  /** Whether the span fits inside the single-room bound. */
  readonly liveAdmitted: boolean;
  readonly summary: string;
}

/**
 * The promotion check Monster Mesh runs before any UI surface displays an
 * acoustic source as live. It folds the Lean contract into an epistemic
 * outcome: a span within the single-room bound is consistent with one room and
 * admitted live (mirroring `live_room_admitted`); a span exceeding the bound can
 * only arise from two convolved rooms — convolution never shortens the tail —
 * and is withheld as a replay (mirroring `replay_rejected_fixture` via
 * `convSpan_ge_left`).
 *
 * @param reflectionSpan  the observed reverberation-tail length in taps (e.g.
 *                        `convSpan(recordRoom, playRoom)` for a suspected replay)
 * @param bound           the single-room bound (defaults to `singleRoomBound`)
 */
export function admitLiveness(
  reflectionSpan: number,
  bound: number = singleRoomBound,
): LivenessAdmission {
  const liveAdmitted = admitsLive(bound, reflectionSpan);

  if (liveAdmitted) {
    return {
      outcome: 'claim_admitted',
      limitation: 'none',
      reflectionSpan,
      singleRoomBound: bound,
      liveAdmitted,
      summary: `reflection span ${reflectionSpan} fits the single-room bound ${bound} — consistent with one room, live`,
    };
  }

  return {
    outcome: 'claim_withheld',
    limitation: 'replay_composition_suspected',
    reflectionSpan,
    singleRoomBound: bound,
    liveAdmitted,
    summary: `reflection span ${reflectionSpan} exceeds the single-room bound ${bound} — only two convolved rooms explain it, replay`,
  };
}

/* ------------------------------------------------------------------ */
/*  Finite certificate                                                */
/* ------------------------------------------------------------------ */

/**
 * A finite certificate that Monster Mesh has all required replay-liveness gates.
 * Mirrors `AcousticReplayLiveness.ReplayLivenessCertificate`.
 */
export interface ReplayLivenessCertificate {
  readonly reflectionSpanModeled: boolean;
  readonly singleRoomBounded: boolean;
  readonly compositionLengthens: boolean;
  readonly replayRejected: boolean;
  readonly livenessAdmitted: boolean;
  readonly hardNegativeSampled: boolean;
}

/** The replay-liveness carrier for Monster Mesh promotion checks.
 *  Mirrors `AcousticReplayLiveness.replayLivenessCertificate`. */
export const replayLivenessCertificate: ReplayLivenessCertificate = {
  reflectionSpanModeled: true,
  singleRoomBounded: true,
  compositionLengthens: true,
  replayRejected: true,
  livenessAdmitted: true,
  hardNegativeSampled: true,
};

/* ------------------------------------------------------------------ */
/*  Concrete fixtures (mirror the Lean fixtures)                      */
/* ------------------------------------------------------------------ */

/** A live room: one room, span 5, well within the single-room bound of 8.
 *  Mirrors `AcousticReplayLiveness.liveRoom`. */
export const liveRoom: RoomEcho = { reflectionSpan: 5 };

/** The room the deepfake was originally recorded in: span 6.
 *  Mirrors `AcousticReplayLiveness.recordRoom`. */
export const recordRoom: RoomEcho = { reflectionSpan: 6 };

/** The room the loudspeaker plays it back into: span 5.
 *  Mirrors `AcousticReplayLiveness.playRoom`. */
export const playRoom: RoomEcho = { reflectionSpan: 5 };

/** The composed replay tail: `convSpan 6 5 = 6 + 5 - 1 = 10`, over the bound 8.
 *  Mirrors `AcousticReplayLiveness.replaySpan`. */
export const replaySpan: number = convSpan(
  recordRoom.reflectionSpan,
  playRoom.reflectionSpan,
);

/**
 * Acoustic Replay Liveness master witness. Bundles the boundary facts the Lean
 * theorem `replay_liveness_master` proves, each recomputed here from first
 * principles (never imported from Lean IR).
 *
 * Mirrors `AcousticReplayLiveness.replay_liveness_master`.
 */
export const ReplayLivenessMaster = {
  /** The live single-room fixture is admitted. */
  liveRoomAdmitted:
    admitLiveness(liveRoom.reflectionSpan).outcome === 'claim_admitted',

  /** The composed replay fixture is rejected (withheld). */
  replayRejected: admitLiveness(replaySpan).outcome === 'claim_withheld',

  /** The concrete replay composition has span 10. */
  replaySpanValue: replaySpan === 10,

  /** Composition strictly lengthens the tail past the live span. */
  replayLengthensPastLive: liveRoom.reflectionSpan < replaySpan,

  /** Composition never shortens the recording-room tail (`convSpan a b >= a`),
   *  with a real playback room (`b >= 1`). */
  convSpanNeverShortensLeft:
    convSpan(recordRoom.reflectionSpan, playRoom.reflectionSpan) >=
    recordRoom.reflectionSpan,

  /** Composition never shortens the playback-room tail (`convSpan a b >= b`),
   *  with a real recording room (`a >= 1`). */
  convSpanNeverShortensRight:
    convSpan(recordRoom.reflectionSpan, playRoom.reflectionSpan) >=
    playRoom.reflectionSpan,

  /** The certificate records replay rejection. */
  certReplayRejected: replayLivenessCertificate.replayRejected,

  /** The certificate records liveness admission. */
  certLivenessAdmitted: replayLivenessCertificate.livenessAdmitted,

  /** The certificate records the same-source/different-liveness hard negative. */
  certHardNegativeSampled: replayLivenessCertificate.hardNegativeSampled,
} as const;
