// @a0n/.ignored_gnosis/distributed-inference-host/src/acoustic-scene-classification.ts
/**
 * # Acoustic Scene Classification — indoor / outdoor / in-vehicle from a signature
 *
 * Mirrors `Gnosis.Mesh.AcousticSceneClassification` from `open-source/gnosis-math`,
 * the Rustic Church carrier for *acoustic scene classification*: from a short
 * signature vector recovered from a window of audio
 * (here `[reverbTail, noiseFloor, lowFreqRumble]`), decide WHICH scene the
 * recording was made in — `indoor`, `outdoor`, or `inVehicle`. Each scene has a
 * prototype signature; the observed signature is compared to all prototypes via
 * the shared list-of-`Nat` `fpDistance`. The Skeptic does NOT simply pick the
 * nearest prototype — a nearest-neighbour vote is happy to commit to a label
 * even when two prototypes are nearly equidistant. Instead a scene label is
 * admitted only when the observation is *confidently* nearest one prototype:
 *
 *  1. **Match gate**: the distance to the claimed prototype clears a tolerance
 *     (`claimedDist <= tol`) — the signature really does look like that scene.
 *  2. **Confident-margin gate** (the hard negative against ambiguity): the
 *     distance to EVERY OTHER prototype exceeds the claimed distance by more
 *     than `sep` (`other > claimedDist + sep`). A signature that sits nearly
 *     equidistant between two prototypes — the genuinely ambiguous case — fails
 *     this gate even when it clears the match tolerance, so it yields no label.
 *
 * The two failure modes are distinct and both refuted:
 *
 *  - A clearly-indoor signature is admitted as `indoor` (small claimed distance,
 *    every other prototype far away) and is NOT admitted as `outdoor` (the
 *    claimed distance to the outdoor prototype blows past `tol`, failing the
 *    match gate).
 *  - An ambiguous signature — nearly equidistant between two prototypes — is
 *    rejected on the MARGIN gate, not the match gate: it clears the (raised)
 *    tolerance but a second prototype sits within `sep` of the claimed distance.
 *
 * ## What is (and is not) carried
 *
 * This module is the runtime promotion check for a scene-label observation. It
 * carries ONLY the finite contract the Lean file proves: the saturating per-band
 * metric (`adist`), the signature distance (`fpDistance`), the confident-margin
 * predicate (`confident`), and their aggregate admission (`admitScene`). Real
 * feature extraction (reverberation-time estimation, spectral noise-floor
 * tracking, low-frequency rumble energy), the actual prototype library, embedding
 * models, and open-set / unknown-scene rejection stay runtime/empirical
 * obligations OUTSIDE this certificate — they are not theorems.
 *
 * The promotion check folds the contract into the same epistemic-outcome
 * vocabulary the sibling mesh carriers already speak
 * (`claim_admitted` / `observation_only` / `claim_withheld`), so a scene-label
 * claim cannot be displayed until it has cleared the Lean-side gates.
 *
 * ## Lean source-of-truth
 *
 * @see open-source/gnosis-math/Gnosis/Mesh/AcousticSceneClassification.lean
 * @see open-source/gnosis-math/Gnosis/Mesh/AcousticSourceCounting.lean (mesh sibling)
 * @see open-source/gnosis-math/Gnosis/Mesh/MicrophoneFingerprint.lean (mesh sibling)
 */

/* ------------------------------------------------------------------ */
/*  Per-band distance and signatures                                  */
/* ------------------------------------------------------------------ */

/**
 * An acoustic-scene signature is a list of per-feature values over a window
 * (here `[reverbTail, noiseFloor, lowFreqRumble]`). A scene prototype is the
 * same shape. Mirrors `AcousticSceneClassification.Signature`.
 */
export type Signature = readonly number[];

/** Absolute difference on `Nat` via saturating subtraction (`|a - b|`). The
 *  per-band signature residual metric, shared with the mesh corridor carriers.
 *  Mirrors `AcousticSceneClassification.adist`. */
export function adist(a: number, b: number): number {
  return Math.max(0, a - b) + Math.max(0, b - a);
}

/**
 * The signature distance: the sum of pointwise `adist` between aligned features.
 * Length-aligned by construction; on a length mismatch the shared elements are
 * summed (mirroring the Lean recursion, which stops at the shorter list).
 * Mirrors `AcousticSceneClassification.fpDistance`.
 */
export function fpDistance(x: Signature, y: Signature): number {
  const n = Math.min(x.length, y.length);
  let sum = 0;
  for (let i = 0; i < n; i++) {
    sum += adist(x[i], y[i]);
  }
  return sum;
}

/* ------------------------------------------------------------------ */
/*  Distances to prototypes                                           */
/* ------------------------------------------------------------------ */

/** A labelled prototype: a scene label and its prototype signature. */
export interface Prototype {
  readonly label: string;
  readonly signature: Signature;
}

/** The distance from an observed signature to each prototype, in order. */
export function distancesToPrototypes(
  signature: Signature,
  prototypes: readonly Prototype[],
): number[] {
  return prototypes.map((p) => fpDistance(signature, p.signature));
}

/* ------------------------------------------------------------------ */
/*  Gate 2: confident margin                                          */
/* ------------------------------------------------------------------ */

/**
 * The confident-margin predicate over the other prototypes' distances: every
 * other distance strictly exceeds the claimed distance plus the separation
 * budget (`other > claimedDist + sep`). A second prototype within the budget
 * marks ambiguity and makes this false.
 * Mirrors `AcousticSceneClassification.confidentB`.
 */
export function confident(
  sep: number,
  claimedDist: number,
  otherDists: readonly number[],
): boolean {
  for (const d of otherDists) {
    if (!(d > claimedDist + sep)) {
      return false;
    }
  }
  return true;
}

/* ------------------------------------------------------------------ */
/*  Runtime promotion check (epistemic outcome)                       */
/* ------------------------------------------------------------------ */

/**
 * Epistemic outcome of a scene-label observation. Identical vocabulary to the
 * sibling mesh carriers, kept as a local literal union so this certificate has
 * no import coupling:
 *
 *  - `claim_admitted`    — the signature clears the match tolerance against the
 *                          claimed prototype AND is confidently separated from
 *                          every other prototype, so the scene label may be
 *                          promoted.
 *  - `observation_only`  — reserved for the degenerate empty-prototype case:
 *                          there is no library to classify against, so the
 *                          observation is reported, never promoted.
 *  - `claim_withheld`    — a gate failed (the match distance exceeds tolerance,
 *                          or a second prototype sits within the margin), so the
 *                          proposed label is refused.
 */
export type SceneEpistemicOutcome =
  | 'claim_admitted'
  | 'observation_only'
  | 'claim_withheld';

/** The reason a scene-label claim was held back from promotion. */
export type SceneLimitationClass =
  | 'match_fail'
  | 'ambiguous_margin'
  | 'no_prototypes'
  | 'none';

export interface SceneAdmission {
  readonly outcome: SceneEpistemicOutcome;
  readonly limitation: SceneLimitationClass;
  /** The claimed scene label (the displayable label when admitted). */
  readonly label: string;
  /** The distance from the signature to the claimed prototype. */
  readonly claimedDistance: number;
  /** The distances to the other prototypes (the displayable margins). */
  readonly otherDistances: number[];
  /** Whether the claimed distance cleared the match tolerance. */
  readonly matched: boolean;
  /** Whether the claim cleared the confident-margin gate. */
  readonly confident: boolean;
  readonly summary: string;
}

export interface SceneAdmitOptions {
  /** Match tolerance: the claimed prototype distance must clear this. */
  readonly tol: number;
  /** Separation budget: every other prototype must exceed the claimed distance
   *  by more than this. */
  readonly sep: number;
}

/**
 * The promotion check Monster Mesh runs before any UI surface displays a
 * scene-label claim. It folds the Lean contract into an epistemic outcome:
 *
 *  - the match gate refutes a wrong-scene label: a claimed distance beyond `tol`
 *    means the signature does not look like that scene, so it is withheld
 *    (`match_fail`, mirroring `not_admits_of_match_fail` / `indoor_not_outdoor`);
 *  - the confident-margin gate refutes ambiguity: a second prototype within the
 *    margin means the label is not confidently nearest, so it is withheld
 *    (`ambiguous_margin`, mirroring `not_admits_of_margin_false` /
 *    `ambiguous_rejected`);
 *  - only a matched, confidently-separated claim is admitted, with the scene
 *    label promoted (mirroring `indoor_admitted`).
 *
 * The empty prototype library is the degenerate observation-only case: there is
 * nothing to classify against.
 *
 * @param signature  the observed scene signature under examination
 * @param prototypes the labelled prototype library
 * @param claimed    the index of the prototype whose label is being asserted
 * @param opts       the match tolerance and separation budget
 */
export function admitScene(
  signature: Signature,
  prototypes: readonly Prototype[],
  claimed: number,
  opts: SceneAdmitOptions,
): SceneAdmission {
  if (prototypes.length === 0) {
    return {
      outcome: 'observation_only',
      limitation: 'no_prototypes',
      label: '',
      claimedDistance: 0,
      otherDistances: [],
      matched: false,
      confident: false,
      summary: 'empty prototype library: no scene to classify against',
    };
  }

  const dists = distancesToPrototypes(signature, prototypes);
  const claimedDistance = dists[claimed];
  const label = prototypes[claimed].label;
  const otherDistances = dists.filter((_d, i) => i !== claimed);
  const matched = claimedDistance <= opts.tol;
  const isConfident = confident(opts.sep, claimedDistance, otherDistances);

  // Match gate: a claimed distance beyond tolerance means the signature does not
  // look like the claimed scene at all — refuse the wrong-scene label.
  if (!matched) {
    return {
      outcome: 'claim_withheld',
      limitation: 'match_fail',
      label,
      claimedDistance,
      otherDistances,
      matched,
      confident: isConfident,
      summary: `match fail: distance ${claimedDistance} to '${label}' exceeds tol ${opts.tol} — signature does not look like this scene`,
    };
  }

  // Confident-margin gate: a second prototype within the margin means the label
  // is not confidently nearest — refuse the ambiguous label.
  if (!isConfident) {
    return {
      outcome: 'claim_withheld',
      limitation: 'ambiguous_margin',
      label,
      claimedDistance,
      otherDistances,
      matched,
      confident: isConfident,
      summary: `ambiguous margin: a second prototype sits within sep ${opts.sep} of the claimed distance ${claimedDistance} — '${label}' is not confidently nearest`,
    };
  }

  return {
    outcome: 'claim_admitted',
    limitation: 'none',
    label,
    claimedDistance,
    otherDistances,
    matched,
    confident: isConfident,
    summary: `matched (distance ${claimedDistance} <= tol ${opts.tol}) and confidently separated (margin > ${opts.sep}) — scene '${label}' certified`,
  };
}

/* ------------------------------------------------------------------ */
/*  Concrete fixtures (mirror the Lean fixtures)                      */
/* ------------------------------------------------------------------ */

/** Indoor prototype: long reverb tail, low noise floor, low rumble.
 *  Mirrors `AcousticSceneClassification.indoor`. */
export const indoor: Prototype = { label: 'indoor', signature: [30, 5, 8] };

/** Outdoor prototype: no reverb, high broadband noise floor, low rumble.
 *  Mirrors `outdoor`. */
export const outdoor: Prototype = { label: 'outdoor', signature: [5, 25, 3] };

/** In-vehicle prototype: some reverb, mid noise, strong low-frequency rumble.
 *  Mirrors `inVehicle`. */
export const inVehicle: Prototype = {
  label: 'inVehicle',
  signature: [10, 15, 40],
};

/** The ordered prototype library (indoor, outdoor, inVehicle). */
export const prototypes: readonly Prototype[] = [indoor, outdoor, inVehicle];

/** Match tolerance: the claimed prototype distance must clear this.
 *  Mirrors `tol`. */
export const tol = 10;

/** Separation budget: every other prototype must exceed the claimed distance by
 *  more than this. Mirrors `sep`. */
export const sep = 8;

/** A clearly-indoor observed signature: very close to the indoor prototype
 *  (distance 4). Mirrors `signatureIndoor`. */
export const signatureIndoor: Signature = [32, 6, 7];

/** An ambiguous observed signature: nearly equidistant between indoor and
 *  outdoor (distance 25 to each), so the confident-margin gate must refuse it
 *  even when the match tolerance is raised enough to admit it.
 *  Mirrors `signatureAmbiguous`. */
export const signatureAmbiguous: Signature = [18, 15, 5];

/** Raised tolerance used only for the ambiguous fixture, chosen so the match
 *  gate passes (25 <= 30) and the rejection lands squarely on the margin gate.
 *  Mirrors `tolAmbig`. */
export const tolAmbig = 30;

/**
 * Acoustic Scene Classification master witness. Bundles the boundary facts the
 * Lean theorem `acoustic_scene_master` proves, each recomputed here from first
 * principles (never imported from Lean IR).
 *
 * Mirrors `AcousticSceneClassification.acoustic_scene_master`.
 */
export const AcousticSceneMaster = {
  /** A signature is at zero distance from itself. */
  fpDistanceSelfZero: fpDistance(signatureIndoor, signatureIndoor) === 0,

  /** The clearly-indoor signature is distance 4 from the indoor prototype. */
  dIndoorIndoor: fpDistance(signatureIndoor, indoor.signature) === 4,

  /** The clearly-indoor signature is distance 50 from the outdoor prototype. */
  dIndoorOutdoor: fpDistance(signatureIndoor, outdoor.signature) === 50,

  /** The clearly-indoor signature is distance 64 from the in-vehicle prototype. */
  dIndoorInVehicle: fpDistance(signatureIndoor, inVehicle.signature) === 64,

  /** The ambiguous signature is distance 25 from the indoor prototype. */
  dAmbigIndoor: fpDistance(signatureAmbiguous, indoor.signature) === 25,

  /** The ambiguous signature is distance 25 from the outdoor prototype — equal to
   *  its indoor distance, the heart of the ambiguity. */
  dAmbigOutdoor: fpDistance(signatureAmbiguous, outdoor.signature) === 25,

  /** The ambiguous signature is distance 43 from the in-vehicle prototype. */
  dAmbigInVehicle: fpDistance(signatureAmbiguous, inVehicle.signature) === 43,

  /** The clearly-indoor signature is admitted as `indoor` (claim 0). */
  indoorAdmitted:
    admitScene(signatureIndoor, prototypes, 0, { tol, sep }).outcome ===
    'claim_admitted',

  /** The clearly-indoor signature is NOT admitted as `outdoor` (claim 1): the
   *  match gate fails (distance 50 > tol 10). */
  indoorNotOutdoor:
    admitScene(signatureIndoor, prototypes, 1, { tol, sep }).limitation ===
    'match_fail',

  /** The ambiguous signature's margin gate fails when claiming `indoor`: the
   *  outdoor distance 25 is not greater than 25 + sep = 33. */
  ambiguousMarginFails:
    confident(sep, 25, [25, 43]) === false,

  /** The ambiguous signature clears the (raised) match gate: 25 <= tolAmbig 30. */
  ambiguousMatchPasses: 25 <= tolAmbig,

  /** The ambiguous signature is rejected as `indoor` on the MARGIN gate, not the
   *  match gate, under the raised tolerance. */
  ambiguousRejected:
    admitScene(signatureAmbiguous, prototypes, 0, { tol: tolAmbig, sep })
      .limitation === 'ambiguous_margin',

  /** The hard negative: scene labels are distinguishable. The indoor signature is
   *  confidently admitted; the ambiguous one — which clears the same kind of
   *  match tolerance — is refused on the margin gate. */
  sceneDistinguishable:
    admitScene(signatureIndoor, prototypes, 0, { tol, sep }).outcome ===
      'claim_admitted' &&
    admitScene(signatureAmbiguous, prototypes, 0, { tol: tolAmbig, sep })
      .outcome === 'claim_withheld',
} as const;
