// @a0n/.ignored_gnosis/distributed-inference-host/src/acoustic-watermark-mesh-identity.ts
/**
 * # Acoustic Watermark Mesh Identity — inaudible provenance, verified acoustically
 *
 * Mirrors `Gnosis.Mesh.AcousticWatermarkMeshIdentity` from
 * `open-source/gnosis-math`, the *inverse* of the single-ear acoustic pose
 * carrier (cf. `./active-acoustic-pose`). There, ambient audio probes the
 * listener and the runtime separates source from pose residue. Here, the played
 * audio CARRIES an inaudibly-embedded mesh-identity token — a protocol69-admitted
 * token — and any mesh node that hears the audio can verify provenance
 * acoustically: "this audio came from our mesh." The watermark rides under the
 * audible content; two clips with identical audible content can differ only in
 * whether (and which) watermark token they carry.
 *
 * ## What is (and is not) carried
 *
 * This module is the runtime promotion check for a mesh-authentic provenance
 * claim. It carries ONLY the finite contract the Lean file proves, two gates:
 *
 *  - **Correlation / support gate** (`DetectedAbove`): the watermark must be
 *    *detected*, i.e. the recovered correlation must clear a threshold. An absent
 *    watermark (noise-level correlation) carries no provenance and is rejected —
 *    noise must not accidentally certify. A genuine embed clears the threshold.
 *  - **Token-admission gate** (`admittedToken`): the recovered token must be an
 *    admitted mesh token. A clean recovery of a token the mesh never issued is a
 *    forgery and is rejected even at perfect correlation.
 *
 * protocol69 note: the real protocol69 predicate (`66^7 ≡ 69` gating) lives in
 * the runtime; here we carry only the finite, decidable contract "the recovered
 * token MATCHES an admitted mesh token", modeled as concrete membership in a
 * small fixed admitted set (`{69, 66, 660}`). `69` is the canonical protocol69
 * admission marker; the others stand in for additional issued tokens.
 *
 * Real spread-spectrum / echo-hiding embedding, psychoacoustic inaudibility,
 * correlation-detector statistics, channel robustness (compression, room,
 * resampling), and the actual protocol69 admission algebra stay runtime/empirical
 * obligations OUTSIDE this certificate — they are not theorems and are not
 * asserted here.
 *
 * The promotion check folds the contract into the same epistemic-outcome
 * vocabulary the sibling mesh carriers already speak
 * (`claim_admitted` / `observation_only` / `claim_withheld`), so a
 * mesh-authentic provenance claim cannot be displayed until it has cleared both
 * Lean-side gates.
 *
 * ## Lean source-of-truth
 *
 * @see open-source/gnosis-math/Gnosis/Mesh/AcousticWatermarkMeshIdentity.lean
 * @see open-source/gnosis-math/Gnosis/Mesh/ActiveAcousticPose.lean (single-ear inverse)
 */

/* ------------------------------------------------------------------ */
/*  Recovered watermark                                               */
/* ------------------------------------------------------------------ */

/**
 * A recovered watermark: the token decoded from the audio, and the detector
 * correlation (how strongly the watermark was detected — `0` = absent / noise,
 * high = a clean recovery). Mirrors `AcousticWatermarkMeshIdentity.RecoveredWatermark`.
 */
export interface RecoveredWatermark {
  readonly token: number;
  readonly correlation: number;
}

/* ------------------------------------------------------------------ */
/*  Gate 2: token-admission (protocol69 finite contract)             */
/* ------------------------------------------------------------------ */

/**
 * Concrete membership test for admitted mesh tokens. The real protocol69
 * predicate (`66^7 ≡ 69`-style gating) lives in the runtime; here we carry only
 * the finite "matches an admitted token" contract against a small fixed set.
 * `69` is the canonical protocol69 admission marker.
 * Mirrors `AcousticWatermarkMeshIdentity.admittedTokenB`.
 */
export function admittedToken(t: number): boolean {
  return t === 69 || t === 66 || t === 660;
}

/* ------------------------------------------------------------------ */
/*  Gate 1: correlation / support                                    */
/* ------------------------------------------------------------------ */

/**
 * The correlation / support gate: the watermark is detected when its recovered
 * correlation clears the threshold. The source-only-rejection analog — an absent
 * or noise-level watermark (low correlation) does not clear it.
 * Mirrors `AcousticWatermarkMeshIdentity.DetectedAbove`.
 */
export function detectedAbove(
  threshold: number,
  wm: RecoveredWatermark,
): boolean {
  return wm.correlation >= threshold;
}

/* ------------------------------------------------------------------ */
/*  Runtime promotion check (epistemic outcome)                       */
/* ------------------------------------------------------------------ */

/**
 * Epistemic outcome of a mesh-authentic provenance observation. Identical
 * vocabulary to the sibling mesh carriers, kept as a local literal union so this
 * certificate has no import coupling:
 *
 *  - `claim_admitted`    — the watermark clears the correlation threshold AND its
 *                          recovered token is an admitted mesh token, so the
 *                          provenance claim may be promoted.
 *  - `observation_only`  — reserved; mesh-authenticity is a hard two-gate
 *                          conjunction and never carries an unquorate middle
 *                          state.
 *  - `claim_withheld`    — a gate failed: a noise-level correlation (the
 *                          watermark was effectively absent) or a token the mesh
 *                          never issued (a forgery), so provenance is refused.
 */
export type WatermarkMeshEpistemicOutcome =
  | 'claim_admitted'
  | 'observation_only'
  | 'claim_withheld';

/** The reason a mesh-authentic provenance claim was held back from promotion. */
export type WatermarkMeshLimitationClass =
  | 'weak_correlation'
  | 'forged_token'
  | 'none';

export interface WatermarkMeshIdentityAdmission {
  readonly outcome: WatermarkMeshEpistemicOutcome;
  readonly limitation: WatermarkMeshLimitationClass;
  /** Whether the correlation gate cleared (correlation >= threshold). */
  readonly correlationCleared: boolean;
  /** Whether the token gate cleared (token is an admitted mesh token). */
  readonly tokenAdmitted: boolean;
  readonly summary: string;
}

export interface WatermarkMeshIdentityAdmitOptions {
  /** The detector correlation a recovery must clear to count as detected. */
  readonly threshold: number;
}

/**
 * The promotion check Monster Mesh runs before any UI surface displays an audio
 * clip as mesh-authentic. It folds the Lean contract into an epistemic outcome:
 *
 *  - the correlation gate refutes an absent / noise-level watermark: a recovery
 *    below threshold carries no provenance and is withheld (mirroring
 *    `low_correlation_rejected`);
 *  - the token gate refutes a forgery: a clean recovery of a token the mesh never
 *    issued is withheld even at perfect correlation (mirroring
 *    `wrong_token_rejected`);
 *  - only a detected-and-admitted recovery is admitted (mirroring
 *    `genuine_admitted` and the `genuine_vs_forged` hard negative).
 *
 * The correlation gate is checked first so a weak forgery reports the
 * (more fundamental) absence of any detectable watermark rather than the token
 * forgery.
 *
 * @param watermark the recovered watermark (decoded token + detector correlation)
 * @param opts      the detector correlation threshold
 */
export function admitMeshAuthentic(
  watermark: RecoveredWatermark,
  opts: WatermarkMeshIdentityAdmitOptions,
): WatermarkMeshIdentityAdmission {
  const correlationCleared = detectedAbove(opts.threshold, watermark);
  const tokenAdmitted = admittedToken(watermark.token);

  // Correlation gate: an absent / noise-level watermark carries no provenance.
  if (!correlationCleared) {
    return {
      outcome: 'claim_withheld',
      limitation: 'weak_correlation',
      correlationCleared,
      tokenAdmitted,
      summary: `weak correlation: recovered correlation ${watermark.correlation} is below detection threshold ${opts.threshold} — no watermark to certify`,
    };
  }

  // Token gate: a clean recovery of a token the mesh never issued is a forgery.
  if (!tokenAdmitted) {
    return {
      outcome: 'claim_withheld',
      limitation: 'forged_token',
      correlationCleared,
      tokenAdmitted,
      summary: `forged token: token ${watermark.token} was never issued by the mesh — strong correlation cannot rescue a forgery`,
    };
  }

  return {
    outcome: 'claim_admitted',
    limitation: 'none',
    correlationCleared,
    tokenAdmitted,
    summary: `mesh-authentic: token ${watermark.token} admitted at correlation ${watermark.correlation} >= threshold ${opts.threshold}`,
  };
}

/* ------------------------------------------------------------------ */
/*  Concrete fixtures (mirror the Lean fixtures)                      */
/* ------------------------------------------------------------------ */

/** The detection threshold used throughout the fixtures. Mirrors the Lean `50`. */
export const watermarkThreshold = 50;

/** A genuine watermarked recovery: an admitted token (69) cleanly detected
 *  (correlation 90). Mirrors `AcousticWatermarkMeshIdentity.genuine`. */
export const genuine: RecoveredWatermark = { token: 69, correlation: 90 };

/** A forged recovery: the SAME strong detector correlation as `genuine`, but the
 *  token (7) was never issued by the mesh. Same audible content, forged
 *  provenance. Mirrors `AcousticWatermarkMeshIdentity.forged`. */
export const forged: RecoveredWatermark = { token: 7, correlation: 90 };

/** A weak recovery: an admitted token (69), but correlation 10 is below
 *  threshold — the watermark was effectively absent / lost in noise.
 *  Mirrors `AcousticWatermarkMeshIdentity.weak`. */
export const weak: RecoveredWatermark = { token: 69, correlation: 10 };

/**
 * Acoustic Watermark Mesh Identity master witness. Bundles the boundary facts the
 * Lean theorem `watermark_mesh_identity_master` proves, each recomputed here from
 * first principles (never imported from Lean IR).
 *
 * Mirrors `AcousticWatermarkMeshIdentity.watermark_mesh_identity_master`.
 */
export const WatermarkMeshIdentityMaster = {
  /** The genuine recovery's token is admitted. */
  genuineTokenAdmitted: admittedToken(genuine.token) === true,

  /** The forged recovery's token is NOT admitted. */
  forgedTokenNotAdmitted: admittedToken(forged.token) === false,

  /** The genuine recovery is admitted as mesh-authentic. */
  genuineAdmitted:
    admitMeshAuthentic(genuine, { threshold: watermarkThreshold }).outcome ===
    'claim_admitted',

  /** The forged recovery is rejected on the token gate (forgery). */
  forgedTokenRejected:
    admitMeshAuthentic(forged, { threshold: watermarkThreshold }).limitation ===
    'forged_token',

  /** The forged recovery is withheld, not admitted. */
  forgedRejected:
    admitMeshAuthentic(forged, { threshold: watermarkThreshold }).outcome ===
    'claim_withheld',

  /** The weak recovery is rejected on the correlation gate (effectively absent). */
  weakCorrelationRejected:
    admitMeshAuthentic(weak, { threshold: watermarkThreshold }).limitation ===
    'weak_correlation',

  /** The weak recovery is withheld, not admitted. */
  weakRejected:
    admitMeshAuthentic(weak, { threshold: watermarkThreshold }).outcome ===
    'claim_withheld',

  /** The standing hard negative: same audible content (same correlation 90), the
   *  genuine provenance admitted and the forged one refused — only the token's
   *  validity distinguishes them. */
  authenticityDistinguished:
    admitMeshAuthentic(genuine, { threshold: watermarkThreshold }).outcome ===
      'claim_admitted' &&
    admitMeshAuthentic(forged, { threshold: watermarkThreshold }).outcome ===
      'claim_withheld',
} as const;
