// @a0n/.ignored_gnosis/distributed-inference-host/src/active-acoustic-pose.ts
/**
 * # Active Acoustic Pose — source rejection, frequency support, hard negatives
 *
 * Mirrors `Gnosis.Mesh.ActiveAcousticPose` from `open-source/gnosis-math`,
 * the Rustic Church carrier for the BGM2Pose-style idea (arXiv:2503.00389):
 * ordinary background audio can act as an active sensing carrier only when the
 * runtime separates source variation from pose-caused acoustic residue.
 *
 * ## What is (and is not) carried
 *
 * This module is the runtime promotion check for an acoustic pose observation.
 * It carries ONLY the finite contract the Lean file proves:
 *
 *  - subtract the tracked source from the recorded bin (`poseResidue`);
 *  - reject source-only bins at positive residue thresholds
 *    (`source_only_not_kept_above_zero`);
 *  - keep bins where the source has support and pose residue clears threshold
 *    (`mixed_pose_kept`);
 *  - treat same-source/different-pose pairs as hard negatives, so source
 *    identity cannot be used as a pose key
 *    (`hard_negative_rejects_source_only_key`).
 *
 * Empirical pose accuracy, STFT/mel correctness, ambisonics physics, and
 * neural contrastive convergence remain runtime/empirical obligations OUTSIDE
 * this certificate — they are not theorems and are not asserted here.
 *
 * The promotion check folds the contract into the same epistemic-outcome
 * vocabulary the mesh carrier already speaks
 * (`claim_admitted` / `claim_withheld` / `observation_only`), so an acoustic
 * observation cannot be displayed as a pose claim until it has cleared the
 * Lean-side residue/support/hard-negative gate.
 *
 * ## Lean source-of-truth
 *
 * @see open-source/gnosis-math/Gnosis/Mesh/ActiveAcousticPose.lean
 * @see open-source/gnosis-math/RUSTIC_CHURCH.md (paper-to-carrier boundary)
 */

/* ------------------------------------------------------------------ */
/*  Frequency bins and acoustic residues                              */
/* ------------------------------------------------------------------ */

/** The proof-facing frequency support used by the active acoustic mesh.
 *  Mirrors `Gnosis.Mesh.ActiveAcousticPose.FrequencyBand`. */
export type FrequencyBand = 'low' | 'mid' | 'high';

/** Stable ordinal used when exporting small certificates to runtime code.
 *  Mirrors `ActiveAcousticPose.bandIndex`. */
export function bandIndex(band: FrequencyBand): number {
  switch (band) {
    case 'low':
      return 0;
    case 'mid':
      return 1;
    case 'high':
      return 2;
  }
}

/**
 * A normalized time-frequency bin: tracked source energy and recorded energy.
 * Energies are non-negative integers (saturating `Nat` arithmetic), matching
 * the Lean `AcousticBin`.
 */
export interface AcousticBin {
  readonly band: FrequencyBand;
  readonly musicEnergy: number;
  readonly recordedEnergy: number;
}

/** Saturating residue after subtracting the tracked source carrier.
 *  Mirrors `ActiveAcousticPose.poseResidue` (`Nat` truncated subtraction). */
export function poseResidue(bin: AcousticBin): number {
  return Math.max(0, bin.recordedEnergy - bin.musicEnergy);
}

/** The sensing carrier is active on a bin when the source has support there.
 *  Mirrors `ActiveAcousticPose.sourceActive`. */
export function sourceActive(bin: AcousticBin): boolean {
  return bin.musicEnergy > 0;
}

/** Frequency-wise support gate: keep only source-supported bins with residue.
 *  Mirrors `ActiveAcousticPose.attentionKeeps`. */
export function attentionKeeps(threshold: number, bin: AcousticBin): boolean {
  return sourceActive(bin) && threshold <= poseResidue(bin);
}

/** A source-only bin contains no pose echo beyond the tracked carrier.
 *  Mirrors `ActiveAcousticPose.sourceOnlyBin`. */
export function sourceOnlyBin(band: FrequencyBand, music: number): AcousticBin {
  return { band, musicEnergy: music, recordedEnergy: music };
}

/** A mixed bin records the tracked carrier plus pose-caused residue.
 *  Mirrors `ActiveAcousticPose.mixedPoseBin`. */
export function mixedPoseBin(
  band: FrequencyBand,
  music: number,
  pose: number,
): AcousticBin {
  return { band, musicEnergy: music, recordedEnergy: music + pose };
}

/**
 * Executable score used by runtime adapters: rejected bins contribute zero.
 * Mirrors `ActiveAcousticPose.attentionScore`. The Lean theorems
 * `source_only_attention_score_zero` and `mixed_pose_attention_score` pin the
 * two boundary values this function must produce.
 */
export function attentionScore(threshold: number, bin: AcousticBin): number {
  return attentionKeeps(threshold, bin) ? poseResidue(bin) : 0;
}

/* ------------------------------------------------------------------ */
/*  Contrastive source rejection                                      */
/* ------------------------------------------------------------------ */

/** A recording links the tracked source id, pose id, and one proof bin.
 *  Mirrors `ActiveAcousticPose.PoseRecording`. */
export interface PoseRecording {
  readonly musicId: number;
  readonly poseId: number;
  readonly bin: AcousticBin;
}

/** Same-source relation for a mini-batch pair. Mirrors `sameMusic`. */
export function sameMusic(left: PoseRecording, right: PoseRecording): boolean {
  return left.musicId === right.musicId;
}

/** Same-pose relation for a mini-batch pair. Mirrors `samePose`. */
export function samePose(left: PoseRecording, right: PoseRecording): boolean {
  return left.poseId === right.poseId;
}

/**
 * BGM-based hard negative: same source carrier, different pose label.
 * Mirrors `ActiveAcousticPose.hardNegative`. A hard negative is the runtime
 * witness that source identity is NOT a usable pose key — the contractual
 * counterpart of `hard_negative_rejects_source_only_key`.
 */
export function hardNegative(
  left: PoseRecording,
  right: PoseRecording,
): boolean {
  return sameMusic(left, right) && !samePose(left, right);
}

/** Does this mini-batch contain at least one same-source/different-pose pair? */
export function hasHardNegative(batch: readonly PoseRecording[]): boolean {
  for (let i = 0; i < batch.length; i++) {
    for (let j = i + 1; j < batch.length; j++) {
      if (hardNegative(batch[i], batch[j]) || hardNegative(batch[j], batch[i])) {
        return true;
      }
    }
  }
  return false;
}

/* ------------------------------------------------------------------ */
/*  Runtime promotion check (epistemic outcome)                       */
/* ------------------------------------------------------------------ */

/**
 * Epistemic outcome of an acoustic pose observation. Identical vocabulary to
 * `MeshCarrierEpistemicOutcome` (`./mesh-carrier-assessment`), kept as a local
 * literal union so this certificate has no import coupling:
 *
 *  - `claim_admitted`    — a source-supported bin clears the residue threshold
 *                          AND a hard negative is present, so the observation
 *                          may be promoted to a pose claim.
 *  - `observation_only`  — a bin clears the gate but no hard negative pins
 *                          source≠pose, so source identity could be doing the
 *                          work; the observation is reported, never promoted.
 *  - `claim_withheld`    — no bin clears the residue gate (e.g. a source-only
 *                          recording), so there is no pose echo to claim.
 */
export type AcousticPoseEpistemicOutcome =
  | 'claim_admitted'
  | 'observation_only'
  | 'claim_withheld';

/** The reason an acoustic pose claim was held back from promotion. */
export type AcousticPoseLimitationClass =
  | 'no_source_supported_residue'
  | 'source_identity_unrefuted'
  | 'none';

export interface AcousticPoseAdmission {
  readonly outcome: AcousticPoseEpistemicOutcome;
  readonly limitation: AcousticPoseLimitationClass;
  /** Bins (by index into the batch) that cleared `attentionKeeps`. */
  readonly keptBins: readonly number[];
  /** Total retained pose residue across kept bins (the displayable signal). */
  readonly retainedResidue: number;
  /** Whether a same-source/different-pose hard negative was present. */
  readonly hardNegativeSampled: boolean;
  readonly summary: string;
}

/**
 * The promotion check Monster Mesh runs before any UI surface displays an
 * acoustic pose observation. It folds the Lean contract into an epistemic
 * outcome: residue/support decides whether there is anything to claim, and the
 * hard negative decides whether the claim is allowed to be a *pose* claim
 * rather than a re-statement of source identity.
 *
 * @param batch     the recordings observed in one mini-batch / window
 * @param threshold the positive residue threshold (must be > 0 to inherit the
 *                  Lean source-only rejection guarantee)
 */
export function admitAcousticPoseObservation(
  batch: readonly PoseRecording[],
  threshold: number,
): AcousticPoseAdmission {
  const keptBins: number[] = [];
  let retainedResidue = 0;
  for (let i = 0; i < batch.length; i++) {
    const score = attentionScore(threshold, batch[i].bin);
    if (score > 0) {
      keptBins.push(i);
      retainedResidue += score;
    }
  }
  const hardNegativeSampled = hasHardNegative(batch);

  // A non-positive threshold cannot inherit the Lean source-only rejection
  // theorem (`source_only_not_kept_above_zero` requires `0 < threshold`), so
  // the gate refuses to admit regardless of the residue.
  const thresholdIsPositive = threshold > 0;

  if (keptBins.length === 0 || !thresholdIsPositive) {
    return {
      outcome: 'claim_withheld',
      limitation: 'no_source_supported_residue',
      keptBins,
      retainedResidue,
      hardNegativeSampled,
      summary: thresholdIsPositive
        ? 'no source-supported bin cleared the residue threshold'
        : `non-positive threshold ${threshold} cannot reject source-only bins`,
    };
  }

  if (!hardNegativeSampled) {
    return {
      outcome: 'observation_only',
      limitation: 'source_identity_unrefuted',
      keptBins,
      retainedResidue,
      hardNegativeSampled,
      summary:
        'residue retained but no same-source/different-pose pair refutes ' +
        'source identity as the pose key',
    };
  }

  return {
    outcome: 'claim_admitted',
    limitation: 'none',
    keptBins,
    retainedResidue,
    hardNegativeSampled,
    summary: `${keptBins.length} bin(s) retained pose residue ${retainedResidue} with source identity refuted`,
  };
}

/* ------------------------------------------------------------------ */
/*  Finite certificate + master witness                               */
/* ------------------------------------------------------------------ */

/**
 * A finite certificate that Monster Mesh has all required acoustic gates.
 * Mirrors `ActiveAcousticPose.ActiveAcousticPoseCertificate`.
 */
export interface ActiveAcousticPoseCertificate {
  readonly originalSourceTracked: boolean;
  readonly residueComputed: boolean;
  readonly frequencySupportApplied: boolean;
  readonly hardNegativeSampled: boolean;
  readonly sourceOnlyRejected: boolean;
  readonly poseEchoRetained: boolean;
}

/** BGM2Pose-inspired finite carrier for Monster Mesh promotion checks.
 *  Mirrors `ActiveAcousticPose.bgm2poseRusticCertificate`. */
export const bgm2poseRusticCertificate: ActiveAcousticPoseCertificate = {
  originalSourceTracked: true,
  residueComputed: true,
  frequencySupportApplied: true,
  hardNegativeSampled: true,
  sourceOnlyRejected: true,
  poseEchoRetained: true,
};

/* ------------------------------------------------------------------ */
/*  Concrete AMPL-shaped fixture (mirrors the Lean fixtures)          */
/* ------------------------------------------------------------------ */

/** Same source (musicId 7), standing pose. Mirrors `ambientStanding`. */
export const ambientStanding: PoseRecording = {
  musicId: 7,
  poseId: 11,
  bin: mixedPoseBin('mid', 5, 3),
};

/** Same source (musicId 7), walking pose. Mirrors `ambientWalking`. */
export const ambientWalking: PoseRecording = {
  musicId: 7,
  poseId: 12,
  bin: mixedPoseBin('mid', 5, 4),
};

/** Same source, no pose echo (source-only bin). Mirrors `ambientNoPose`. */
export const ambientNoPose: PoseRecording = {
  musicId: 7,
  poseId: 11,
  bin: sourceOnlyBin('mid', 5),
};

/**
 * Active Acoustic Pose master witness. Bundles the boundary facts the Lean
 * theorem `active_acoustic_pose_mesh_witness` proves, each recomputed here
 * from first principles (never imported from Lean IR).
 *
 * Mirrors `ActiveAcousticPose.active_acoustic_pose_mesh_witness`.
 */
export const ActiveAcousticPoseMaster = {
  /** The same-music/different-pose pair is a hard negative. */
  ambientPairIsHardNegative: hardNegative(ambientStanding, ambientWalking),

  /** The no-pose fixture scores zero at a positive residue threshold. */
  ambientNoPoseRejected: attentionScore(1, ambientNoPose.bin) === 0,

  /** The standing fixture keeps its finite pose residue (3). */
  ambientStandingResidueKept: attentionScore(1, ambientStanding.bin) === 3,

  /** The walking fixture keeps its finite pose residue (4). */
  ambientWalkingResidueKept: attentionScore(1, ambientWalking.bin) === 4,

  /** The certificate records source-only rejection. */
  sourceOnlyRejected: bgm2poseRusticCertificate.sourceOnlyRejected,

  /** The certificate records pose-echo retention. */
  poseEchoRetained: bgm2poseRusticCertificate.poseEchoRetained,
} as const;
