// @a0n/.ignored_gnosis/distributed-inference-host/src/alpha-jump.ts
/**
 * alpha-jump.ts — Death #4 (Octonion Fano-plane routing) dispatcher.
 *
 * SCOPE
 * -----
 * Workspace-shared module. Encodes a residual-seed frame (Death #4) and
 * dispatches it on a `FlowTransport` (Death #3 substrate). Framework-agnostic
 * — does NOT know about Cloudflare Workers, fetch, Durable Objects, or
 * specific worker URLs. The caller (a coordinator or bench script) supplies
 * an already-connected `FlowTransport` and a destination URL string for
 * routing/logging.
 *
 * REFERENCES
 * ----------
 * - `open-source/gnosis/distributed-inference/OCTONION_ROUTING_PLAN.md` §3
 *   (Fano-plane assignment), §4 (Pair X — the skip-attention shortcut),
 *   §5 (α-teleportation jump primitive).
 * - `open-source/gnosis-math/Gnosis/FanoOctonionNonAssoc.lean` for the
 *   canonical 7-line table that this module encodes.
 *
 * SHADOW MODE
 * -----------
 * v1 is shadow-mode (config.shadow=true). The dispatcher fires the jumped
 * frame BUT the receiving worker (`/residual-seed`) does NOT yet override
 * the canonical chain residual. The bench compares fingerprints; only after
 * a measurement window without divergence does shadow=false take effect.
 *
 * LIVE MODE GATE (3-state truth table)
 * ------------------------------------
 * The receiving worker (residual-seed-handler.ts) emits a JSON ack with
 * `shadow_mode`, `fully_resolved`, and (in live + partial) `awaiting_resolution`.
 * `shouldSkipCanonicalHop(ack)` encodes the gate the coordinator must apply
 * before issuing the next layer's `/split-a` / `forward` HTTP call:
 *
 *   shadow_mode | fully_resolved | awaiting_resolution | skipCanonical
 *   ------------+----------------+---------------------+---------------
 *   true        | (any)          | (any)               | false   (canonical wins, always)
 *   false       | true           | absent              | TRUE    (Fano sum substitutes)
 *   false       | false          | true                | false   (incomplete bitset; wait it out)
 *
 * Anything outside this table — missing fields, ack fetch failure, mid-stream
 * race where bitset goes partial again — degrades to `false` (canonical wins).
 * The substrate-level guarantee from OCTONION_ROUTING_PLAN.md §5 ("the floor
 * is no worse than today") rides on this default.
 *
 * STREAM ID + FLAGS LAYOUT
 * ------------------------
 * One frame per jump:
 *   streamId  STREAM_RESIDUAL_SEED_BASE + dstLayer  (0xAF60..0xAF6F)
 *   sequence  tokenSeq (u32 monotonic per token)
 *   flags     low nibble: orientation (ORIENTATION_POS | ORIENTATION_NEG)
 *             high nibble: Fano line index 0..6 (encoded in flags >> 4)
 *   payload   8-byte (startPos u32LE | seqLen u32LE) + f32 body
 *
 * The line-index packing is symmetric on the wire — see
 * `apps/worker-node1/src/residual-seed-handler.ts` for the decoder.
 */

// TODO(user): wire `@a0n/aeon` into this package's package.json `dependencies`
// so the workspace resolver can pick it up. Currently `distributed-inference-host`
// has no @a0n/aeon dep — adding `"@a0n/aeon": "workspace:*"` to dependencies
// + running `pnpm install` makes this typecheck. Additionally, `@a0n/aeon`'s
// tsup config has `dts: false`, so the dist lacks `.d.ts` files; either:
//   (a) flip tsup `dts: true` in open-source/aeon/tsup.config.ts, or
//   (b) add a path mapping in this package's tsconfig.json:
//         "paths": { "@a0n/aeon/flow": ["../../aeon/src/flow/index.ts"] }
//   (b) is non-invasive and matches how apps/worker-node1 resolves the type.
import {
  FlowCodec,
  HEADER_SIZE,
  STREAM_RESIDUAL_SEED_BASE,
  ORIENTATION_POS,
  ORIENTATION_NEG,
  encodeInferencePayload,
  type FlowTransport,
} from '@a0n/aeon/flow';

// ── Fano lines (canonical from FanoOctonionNonAssoc.lean) ────────────────

/**
 * The seven canonical Fano-plane lines from `FanoOctonionNonAssoc.lean:99-160`,
 * with mesh-role names per OCTONION_ROUTING_PLAN.md §3.
 *
 *   e1 = cobordism embed mesh
 *   e2 = tri-A (ATTN) ring
 *   e3 = tri-G (GATE-UP) ring
 *   e4 = tri-D (DOWN) ring
 *   e5 = cobordism lm_head mesh
 *   e6 = layer-warden (residual sidecar / KV cache)
 *   e7 = cobordism-warden (vocab partition oracle)
 *
 * Indices match the bit position used in the residual-seed handler's
 * bitset (line N → bit N).
 */
export enum FanoLine {
  /** (e1,e2,e3) — Embed → ATTN → GATE-UP. Warm token onset. */
  E1_E2_E3 = 0,
  /** (e1,e4,e5) — Embed → DOWN → lm_head. THE skip-attention shortcut (Pair X). */
  E1_E4_E5 = 1,
  /** (e1,e7,e6) — Embed → cobordism-warden → layer-warden. Sidecar. */
  E1_E7_E6 = 2,
  /** (e2,e4,e6) — ATTN → DOWN → layer-warden. Pair Y attn-contribution line. */
  E2_E4_E6 = 3,
  /** (e2,e5,e7) — ATTN → lm_head → cobordism-warden. Pair Z early-exit probe. */
  E2_E5_E7 = 4,
  /** (e3,e4,e7) — GATE-UP → DOWN → cobordism-warden. FFN-contribution line. */
  E3_E4_E7 = 5,
  /** (e3,e6,e5) — GATE-UP → layer-warden → lm_head. Mid-FFN sidecar (Amplituhedron). */
  E3_E6_E5 = 6,
}

/** All seven lines, in canonical order. */
export const FANO_LINES: readonly FanoLine[] = [
  FanoLine.E1_E2_E3,
  FanoLine.E1_E4_E5,
  FanoLine.E1_E7_E6,
  FanoLine.E2_E4_E6,
  FanoLine.E2_E5_E7,
  FanoLine.E3_E4_E7,
  FanoLine.E3_E6_E5,
] as const;

function encodeResidualSeedPayload(startPos: number, seqLen: number, body: Uint8Array): Uint8Array {
  const payload = new Uint8Array(8 + body.byteLength);
  const view = new DataView(payload.buffer);
  view.setUint32(0, startPos >>> 0, true);
  view.setUint32(4, seqLen >>> 0, true);
  payload.set(body, 8);
  return encodeInferencePayload(payload);
}

/** Human-readable label for each line. */
export const FANO_LINE_LABEL: Record<FanoLine, string> = {
  [FanoLine.E1_E2_E3]: '(e1,e2,e3) embed→ATTN→GATE-UP',
  [FanoLine.E1_E4_E5]: '(e1,e4,e5) embed→DOWN→lm_head [Pair X]',
  [FanoLine.E1_E7_E6]: '(e1,e7,e6) embed→cob-warden→layer-warden',
  [FanoLine.E2_E4_E6]: '(e2,e4,e6) ATTN→DOWN→layer-warden [Pair Y]',
  [FanoLine.E2_E5_E7]: '(e2,e5,e7) ATTN→lm_head→cob-warden [Pair Z]',
  [FanoLine.E3_E4_E7]: '(e3,e4,e7) GATE-UP→DOWN→cob-warden',
  [FanoLine.E3_E6_E5]: '(e3,e6,e5) GATE-UP→layer-warden→lm_head',
};

// ── Pair X (and friends) — destination/orientation table ─────────────────

/**
 * For each line, derive `(dstLayer, orientation)` from `srcLayer`.
 *
 * Pair X (line E1_E4_E5) is the skip-attention shortcut: the residual that
 * was already known *before* layer `srcLayer` started can teleport directly
 * to layer `srcLayer + 1`'s ATTN input port. Plan §4: "ride that buffer down
 * line (e₁, e₄, e₅) directly to tri-a-(NN+1)'s residual seed port."
 *
 * `fano_non_associative` (FanoOctonionNonAssoc.lean:219) tells us the same
 * destination index can carry either sign depending on the AST grouping. We
 * fix one canonical sign per line; the receiving worker honors whatever the
 * dispatcher set.
 */
interface JumpDestination {
  dstLayer: number;
  orientation: 'pos' | 'neg';
}

// ── Receiver ack shape + live-mode gate ──────────────────────────────────

/**
 * Subset of the JSON body the worker's `/residual-seed` handler returns
 * (see `apps/worker-node1/src/residual-seed-handler.ts:396-408`). Only the
 * fields the live-mode gate consults are typed; the receiver also emits
 * `ack`, `line`, `dst_layer`, `token_seq`, `bitset`, `sign`, `bytes`, which
 * are diagnostic and ignored here.
 *
 * `awaiting_resolution` is OPTIONAL on the wire — the receiver only sets it
 * to `true` when in live mode AND the bitset is still partial. Shadow mode
 * and the fully-resolved live-mode case both omit the field. The gate
 * treats absent === false.
 */
export interface ResidualSeedAck {
  shadow_mode: boolean;
  fully_resolved: boolean;
  awaiting_resolution?: boolean;
}

/**
 * Decide whether the coordinator can SKIP the canonical-chain hop for this
 * residual contribution. Pure function so callers can unit-test the gate
 * without standing up a transport or HTTP probe.
 *
 * Truth table (reproduced from the module header):
 *
 *   shadow_mode | fully_resolved | awaiting_resolution | result
 *   ------------+----------------+---------------------+--------
 *   true        | (any)          | (any)               | false
 *   false       | true           | absent / false      | TRUE
 *   false       | false          | true                | false
 *
 * Any other shape — missing fields, malformed ack, or a contradictory
 * combination like (live + fully_resolved + awaiting_resolution=true) —
 * collapses to `false`. Treat the canonical chain as the safe ground state;
 * the Fano sum is the optimization, never the fallback.
 */
export function shouldSkipCanonicalHop(ack: ResidualSeedAck | null | undefined): boolean {
  if (!ack || typeof ack !== 'object') return false;
  if (ack.shadow_mode !== false) return false;
  if (ack.fully_resolved !== true) return false;
  // Race-condition safety: if the receiver flipped the bit back to
  // partial mid-stream, the awaiting_resolution flag will be true and
  // we MUST fall back. (Today the receiver never sets both flags
  // together, but we don't trust it — the floor is the canonical chain.)
  if (ack.awaiting_resolution === true) return false;
  return true;
}

// ── Dispatcher ────────────────────────────────────────────────────────────

export interface AlphaJumpConfig {
  /**
   * v1: when `shadow=true` the dispatcher fires the frame but the receiver
   * does NOT replace the canonical residual. Required for the first
   * deployment per OCTONION_ROUTING_PLAN.md §5.
   */
  shadow: boolean;
}

export interface AlphaJumpFrameOpts {
  line: FanoLine;
  sign: 'pos' | 'neg';
  payload: Float32Array;
  srcLayer: number;
  dstLayer: number;
  tokenSeq: number;
  /**
   * Identifier of the destination worker. Used for logging and (in non-shadow
   * configurations) for the transport's routing decisions. Pure-string —
   * the dispatcher does not parse it.
   */
  dstWorkerUrl: string;
  /**
   * Position of the first token in this batch. Travels in the 8-byte payload
   * prefix; defaults to 0 (single-step decode case).
   */
  startPos?: number;
  /**
   * Number of tokens in this batch. Travels in the 8-byte payload prefix;
   * defaults to 1 (single-step decode).
   */
  seqLen?: number;
}

/**
 * Encodes one Fano-line residual contribution as a FlowCodec frame and
 * dispatches it on the supplied transport.
 *
 * The dispatcher is intentionally stateless — slot accumulation lives on the
 * receiver (`apps/worker-node1/src/residual-seed-handler.ts`). The caller
 * holds the FlowTransport and is responsible for its lifecycle.
 */
export class AlphaJumpDispatcher {
  private readonly transport: FlowTransport;
  private readonly shadow: boolean;
  private codec: FlowCodec | null = null;
  private codecPromise: Promise<FlowCodec> | null = null;

  constructor(transport: FlowTransport, config: AlphaJumpConfig) {
    this.transport = transport;
    this.shadow = config.shadow;
  }

  /** Whether v1 shadow-mode is in effect. Diagnostic only. */
  isShadow(): boolean {
    return this.shadow;
  }

  /** Lazy-create a JS-only codec (10-byte header — no need for WASM). */
  private async getCodec(): Promise<FlowCodec> {
    if (this.codec) return this.codec;
    if (!this.codecPromise) {
      this.codecPromise = FlowCodec.create({ wasmMode: 'off' });
    }
    this.codec = await this.codecPromise;
    return this.codec;
  }

  /**
   * Dispatch one residual contribution along `line` to `dstLayer`.
   *
   * Returns when the frame has been handed to `transport.send()`. The
   * underlying transport's delivery semantics decide when (and whether)
   * the receiver actually sees the bytes — see Death #3 §1.
   */
  async dispatchSeed(opts: AlphaJumpFrameOpts): Promise<void> {
    const startPos = opts.startPos ?? 0;
    const seqLen = opts.seqLen ?? 1;

    // f32 → bytes (little-endian). We copy because the caller's buffer may
    // be reused once dispatchSeed returns, and the codec keeps a zerocopy
    // view into the payload during encode.
    const f32 = opts.payload;
    const body = new Uint8Array(f32.byteLength);
    {
      const dv = new DataView(body.buffer);
      for (let i = 0; i < f32.length; i++) {
        dv.setFloat32(i * 4, f32[i] ?? 0, /* littleEndian */ true);
      }
    }
    const payload = encodeResidualSeedPayload(startPos, seqLen, body);

    const orientation =
      opts.sign === 'neg' ? ORIENTATION_NEG : ORIENTATION_POS;
    // Pack: high nibble = line index, low nibble = orientation.
    const flags = ((opts.line & 0x07) << 4) | (orientation & 0x0f);

    const streamId = STREAM_RESIDUAL_SEED_BASE + (opts.dstLayer & 0x0f);
    if (opts.dstLayer < 0 || opts.dstLayer > 0x0f) {
      // The 0xAF60..0xAF6F band reserves 16 slots — one per layer in the
      // common 28-layer model after wrapping. Plan §5 keeps the band tight;
      // multi-layer models address this by using `dstLayer & 0x0f` and
      // disambiguating via the `seq` upper bits if needed.
      throw new RangeError(
        `AlphaJumpDispatcher: dstLayer ${opts.dstLayer} outside [0..15]; ` +
          `the residual-seed stream-id band 0xAF60..0xAF6F is exhausted.`,
      );
    }

    const codec = await this.getCodec();
    const frame = codec.encode({
      streamId,
      sequence: opts.tokenSeq >>> 0,
      flags: flags & 0xff,
      payload,
    });

    if (frame.byteLength < HEADER_SIZE) {
      throw new Error(
        `AlphaJumpDispatcher: encoded frame is shorter than header (${frame.byteLength} < ${HEADER_SIZE})`,
      );
    }

    this.transport.send(frame);

    // Shadow-mode is observed at the *receiver*; the dispatcher fires
    // identically. If we needed to gate the dispatch itself on shadow we
    // would do so here — but the bench wants the wire bytes regardless.
    if (this.shadow) {
      // No-op marker; left as a hook for future log/trace integration so
      // the call site documents the mode. Avoids a dead branch warning.
      void opts.dstWorkerUrl;
    }
  }

  /**
   * Dispatch one residual contribution via HTTP POST `/residual-seed`,
   * read the receiver's JSON ack, and decide whether the coordinator can
   * skip the next canonical-chain hop.
   *
   * This is the live-mode entry point. The transport `send()` substrate
   * is fire-and-forget (Death #3 WebSocket), so it cannot inform the
   * coordinator that the bitset is fully resolved. The HTTP path is the
   * one that lets the dispatcher act on `shouldSkipCanonicalHop`.
   *
   * Returns `{ ack, skipCanonical }`:
   *   - `ack` is the parsed JSON body (or `null` on any non-2xx / parse
   *     error / network error — the gate then degrades to false).
   *   - `skipCanonical` is the gate decision per `shouldSkipCanonicalHop`.
   *
   * Logs CONSPICUOUSLY when the substitute fires — this is the moment the
   * Fano sum replaces the canonical hop and any silent regression here is
   * a model-quality regression downstream.
   *
   * Caller contract: when `skipCanonical === true` the caller MUST NOT
   * issue the next layer's `/split-a` (or `forward`) HTTP call; the Fano
   * sum the receiver has already accumulated IS the substitute residual
   * for that destination port. When `false`, fall through to the canonical
   * chain unchanged.
   */
  async dispatchSeedHttp(
    opts: AlphaJumpFrameOpts,
    fetchImpl: typeof fetch = fetch,
    initOverrides?: Partial<RequestInit>,
  ): Promise<{ ack: ResidualSeedAck | null; skipCanonical: boolean }> {
    const startPos = opts.startPos ?? 0;
    const seqLen = opts.seqLen ?? 1;

    const f32 = opts.payload;
    const body = new Uint8Array(f32.byteLength);
    {
      const dv = new DataView(body.buffer);
      for (let i = 0; i < f32.length; i++) {
        dv.setFloat32(i * 4, f32[i] ?? 0, /* littleEndian */ true);
      }
    }
    const payload = encodeResidualSeedPayload(startPos, seqLen, body);

    const orientation =
      opts.sign === 'neg' ? ORIENTATION_NEG : ORIENTATION_POS;
    const flags = ((opts.line & 0x07) << 4) | (orientation & 0x0f);

    const streamId = STREAM_RESIDUAL_SEED_BASE + (opts.dstLayer & 0x0f);
    if (opts.dstLayer < 0 || opts.dstLayer > 0x0f) {
      throw new RangeError(
        `AlphaJumpDispatcher: dstLayer ${opts.dstLayer} outside [0..15]; ` +
          `the residual-seed stream-id band 0xAF60..0xAF6F is exhausted.`,
      );
    }

    const codec = await this.getCodec();
    const frame = codec.encode({
      streamId,
      sequence: opts.tokenSeq >>> 0,
      flags: flags & 0xff,
      payload,
    });

    if (frame.byteLength < HEADER_SIZE) {
      throw new Error(
        `AlphaJumpDispatcher: encoded frame is shorter than header (${frame.byteLength} < ${HEADER_SIZE})`,
      );
    }

    // Build the POST. We accept an `initOverrides` hook so the caller can
    // pass auth (Authorization: Bearer ...), abort signals, or custom
    // content-type without us hard-coding any of it.
    const url = `${stripTrailingSlash(opts.dstWorkerUrl)}/residual-seed`;
    const baseHeaders: Record<string, string> = {
      'Content-Type': 'application/octet-stream',
    };
    const init: RequestInit = {
      method: 'POST',
      headers: { ...baseHeaders, ...(initOverrides?.headers as Record<string, string> | undefined) },
      body: frame as unknown as BodyInit,
      ...stripHeadersAndBody(initOverrides),
    };

    let ack: ResidualSeedAck | null = null;
    try {
      const res = await fetchImpl(url, init);
      if (!res.ok) {
        // Non-2xx: the receiver may or may not have accumulated the seed.
        // Either way we cannot trust the bitset state; fall back to canonical.
        ack = null;
      } else {
        const txt = await res.text();
        try {
          ack = JSON.parse(txt) as ResidualSeedAck;
        } catch {
          ack = null;
        }
      }
    } catch {
      // Network error / aborted / DNS — degrade to canonical. The Fano sum
      // path is an optimization; the canonical chain is the floor.
      ack = null;
    }

    const skipCanonical = shouldSkipCanonicalHop(ack);

    if (skipCanonical) {
      // CONSPICUOUS LOG: this is the moment the Fano sum replaces the
      // canonical hop. If the substitute is wrong, the model output
      // regresses silently downstream. Make sure the line is grep-able
      // in operator log triage.
      console.log(
        `[AlphaJump] LIVE: skipping canonical hop, Fano sum substituted ` +
          `(line=${FANO_LINE_LABEL[opts.line] ?? opts.line} ` +
          `dst_layer=${opts.dstLayer} token_seq=${opts.tokenSeq >>> 0} ` +
          `dst=${opts.dstWorkerUrl})`,
      );
    }

    return { ack, skipCanonical };
  }

  /**
   * Static destination/orientation table.
   *
   * For a given Fano line and source layer, returns the destination layer
   * and the canonical sign for the contribution riding that line. The
   * mesh interpretations are documented on the FanoLine enum.
   *
   * Pair X (E1_E4_E5): srcLayer's pre-attention residual teleports to
   * (srcLayer + 1)'s ATTN input. Sign is positive — the residual stream
   * is additive and the contribution is x_l itself.
   *
   * Pair Y (E2_E4_E6): srcLayer's ATTN output teleports to layer-warden
   * for KV write-back, off the hot path. dstLayer mirrors srcLayer (KV is
   * keyed by srcLayer).
   *
   * Pair Z (E2_E5_E7): srcLayer's ATTN output teleports to lm_head for a
   * speculative early-exit probe. dstLayer is the final layer; sign is
   * positive (additive logit contribution).
   *
   * Other lines: documented in the body.
   */
  static computeJumpDestination(
    line: FanoLine,
    srcLayer: number,
  ): JumpDestination {
    switch (line) {
      case FanoLine.E1_E4_E5:
        // Pair X — skip-attention shortcut. x_l → tri-a-(l+1) input.
        return { dstLayer: srcLayer + 1, orientation: 'pos' };

      case FanoLine.E2_E4_E6:
        // Pair Y — attn output → layer-warden KV. dstLayer == srcLayer.
        return { dstLayer: srcLayer, orientation: 'pos' };

      case FanoLine.E2_E5_E7:
        // Pair Z — attn → lm_head speculative. dstLayer encoded as srcLayer
        // (the lm_head consumer maps it to its own slot table).
        return { dstLayer: srcLayer, orientation: 'pos' };

      case FanoLine.E1_E2_E3:
        // Warm token onset — embedding → ATTN of layer 0. Source must be 0.
        return { dstLayer: srcLayer, orientation: 'pos' };

      case FanoLine.E1_E7_E6:
        // Embed → cobordism-warden → layer-warden. Sidecar; same layer.
        return { dstLayer: srcLayer, orientation: 'pos' };

      case FanoLine.E3_E4_E7:
        // FFN contribution → DOWN → cob-warden. Same layer.
        return { dstLayer: srcLayer, orientation: 'pos' };

      case FanoLine.E3_E6_E5:
        // Mid-FFN sidecar — Amplituhedron capture point. Same layer.
        // fano_non_associative tags this with the opposite-sign variant
        // when consumed by Amplituhedron replay; v1 keeps it positive
        // and lets Death #2 flip if/when needed.
        return { dstLayer: srcLayer, orientation: 'pos' };

      default: {
        // Exhaustiveness — the switch is total over the seven enum values.
        const _exhaustive: never = line;
        return _exhaustive;
      }
    }
  }
}

// ── Helpers (private) ─────────────────────────────────────────────────────

function stripTrailingSlash(url: string): string {
  return url.endsWith('/') ? url.slice(0, -1) : url;
}

/**
 * Pull `headers` and `body` off the override bag so we can layer them
 * ourselves (we set `body` from the encoded frame and merge `headers` so
 * the caller cannot accidentally erase Content-Type). Everything else
 * (signal, credentials, mode, redirect, ...) passes through untouched.
 */
function stripHeadersAndBody(
  init: Partial<RequestInit> | undefined,
): Omit<RequestInit, 'headers' | 'body' | 'method'> {
  if (!init) return {};
  const { headers: _h, body: _b, method: _m, ...rest } = init;
  void _h;
  void _b;
  void _m;
  return rest;
}
