// @a0n/.ignored_gnosis/distributed-inference-host/src/answer-cache.ts
/**
 * Local EXACT-MATCH answer cache — in-process (per warm instance), safe by
 * construction.
 *
 * Unlike the skymesh global cache (model-agnostic, token-hash, semantic-adjacent)
 * and the sovereign sieve (paraphrase/semantic), this cache is a plain
 * IDENTICAL-REQUEST → IDENTICAL-RESPONSE memo: a sha256 over every CACHE-SALIENT
 * request field (model + full messages array + max_tokens + sampling params +
 * prosody-on) keys the completion the deterministic (greedy) decode produced.
 * Two requests with the same key are guaranteed to (re)produce the same tokens,
 * so replaying the stored answer is byte-honest — it is exactly what a fresh
 * generate would return (fresh id/created, reused content + token counts).
 *
 * Store: an insertion-ordered Map used as an LRU (cap by entry count AND bytes,
 * evict oldest). Optional disk mirror (MOONSHINE_ANSWER_CACHE_DIR) survives within
 * the container across the request stream — Cloud Run disk is ephemeral per
 * instance, so cross-instance persistence is a follow-up (skymesh/R2), out of
 * scope here. DEFAULT-ON / opt-out (MOONSHINE_ANSWER_CACHE=0).
 */

import { createHash } from 'node:crypto';
import {
  existsSync,
  mkdirSync,
  readFileSync,
  writeFileSync,
} from 'node:fs';
import { join } from 'node:path';

/** ON by default; off only when MOONSHINE_ANSWER_CACHE is '0'/'false'/'off'. */
export function answerCacheEnabled(): boolean {
  const raw = process.env['MOONSHINE_ANSWER_CACHE'];
  if (raw === undefined) return true;
  const s = raw.trim().toLowerCase();
  return s !== '0' && s !== 'false' && s !== 'off';
}

const MAX_ENTRIES = Number(process.env['MOONSHINE_ANSWER_CACHE_MAX'] ?? 512);
const MAX_BYTES = Number(
  process.env['MOONSHINE_ANSWER_CACHE_MAX_BYTES'] ?? 256 * 1024 * 1024,
);
const DISK_DIR = process.env['MOONSHINE_ANSWER_CACHE_DIR'];

/** The salient fields that determine the completion. Order is fixed + explicit. */
export interface AnswerCacheKeyInput {
  model: string;
  /** role + normalized text content, in message order. */
  messages: ReadonlyArray<{ role: string; content: string }>;
  /** Normalized (never the raw `?? default` gap) — absent == default collide. */
  maxTokens: number;
  temperature: number | null;
  topP: number | null;
  /** Stop sequences, verbatim (order-significant to the decoder). */
  stop: string[] | string | null;
  /** response_format.type (text | json_object) — conditions the answer shape. */
  responseFormat: string | null;
  /**
   * Whether terminal-prosody biasing was ON for this generate. Prosody changes
   * the DECODED tokens (bias), not just the attached `song`, so it is salient.
   */
  prosodyOn: boolean;
}

export interface AnswerCacheEntry {
  /** The assistant text (already stripped/trimmed, exactly as served). */
  content: string;
  /** Decoded completion token count (for the usage block). */
  completionTokens: number;
  /** Prompt token count (for the usage block). */
  promptTokens: number;
  /** Provenance only — NOT part of the key beyond `model` above. */
  model: string;
}

/**
 * sha256 (collision-safe — this is a correctness-critical EXACT cache, so a
 * fp48/FNV bucket would risk serving a wrong answer on a collision) over a
 * canonical JSON of the salient fields in a fixed order.
 */
export function answerCacheKey(input: AnswerCacheKeyInput): string {
  const canonical = JSON.stringify([
    'answer-cache/v1',
    input.model,
    input.messages.map((m) => [m.role, m.content]),
    input.maxTokens,
    input.temperature,
    input.topP,
    input.stop,
    input.responseFormat,
    input.prosodyOn,
  ]);
  return createHash('sha256').update(canonical).digest('hex');
}

// ── In-process LRU (insertion-ordered Map) ──

interface StoredEntry extends AnswerCacheEntry {
  bytes: number;
}
const mem = new Map<string, StoredEntry>();
let memBytes = 0;

function entryBytes(e: AnswerCacheEntry): number {
  // UTF-16-ish upper bound on the retained string plus small numeric overhead.
  return e.content.length * 2 + 64;
}

function evictToBounds(): void {
  while (mem.size > MAX_ENTRIES || memBytes > MAX_BYTES) {
    const oldest = mem.keys().next();
    if (oldest.done) break;
    const victim = mem.get(oldest.value);
    mem.delete(oldest.value);
    if (victim) memBytes -= victim.bytes;
  }
}

function diskPath(key: string): string | null {
  if (!DISK_DIR) return null;
  return join(DISK_DIR, `${key}.json`);
}

let diskDirReady = false;
function ensureDiskDir(): boolean {
  if (!DISK_DIR) return false;
  if (diskDirReady) return true;
  try {
    if (!existsSync(DISK_DIR)) mkdirSync(DISK_DIR, { recursive: true });
    diskDirReady = true;
    return true;
  } catch {
    return false;
  }
}

/**
 * Look up an exact-match completion. Checks the in-process LRU first; on a
 * miss, hydrates from the optional disk mirror (and promotes into memory).
 * Returns `null` on any miss or read error (fail-safe → the caller decodes).
 */
export function answerCacheGet(key: string): AnswerCacheEntry | null {
  const hit = mem.get(key);
  if (hit) {
    // Refresh recency (delete + re-insert moves it to the newest slot).
    mem.delete(key);
    mem.set(key, hit);
    return { content: hit.content, completionTokens: hit.completionTokens, promptTokens: hit.promptTokens, model: hit.model };
  }
  const path = diskPath(key);
  if (path && existsSync(path)) {
    try {
      const parsed = JSON.parse(readFileSync(path, 'utf8')) as AnswerCacheEntry;
      if (
        parsed &&
        typeof parsed.content === 'string' &&
        parsed.content.length > 0 &&
        typeof parsed.completionTokens === 'number' &&
        typeof parsed.promptTokens === 'number'
      ) {
        answerCachePut(key, parsed, /* persist */ false); // hydrate memory only
        return parsed;
      }
    } catch {
      /* corrupt/partial file → treat as miss */
    }
  }
  return null;
}

/**
 * Store an exact-match completion. Never throws. Empty content is not stored
 * (the caller also gates on errors / non-cacheable requests). `persist`
 * controls whether the disk mirror is written (false when hydrating FROM disk).
 */
export function answerCachePut(
  key: string,
  entry: AnswerCacheEntry,
  persist = true,
): void {
  if (!entry.content) return;
  const stored: StoredEntry = { ...entry, bytes: entryBytes(entry) };
  const prev = mem.get(key);
  if (prev) memBytes -= prev.bytes;
  mem.set(key, stored);
  memBytes += stored.bytes;
  evictToBounds();
  if (persist && ensureDiskDir()) {
    const path = diskPath(key);
    if (path) {
      try {
        writeFileSync(
          path,
          JSON.stringify({
            content: entry.content,
            completionTokens: entry.completionTokens,
            promptTokens: entry.promptTokens,
            model: entry.model,
          }),
        );
      } catch {
        /* disk full / read-only → memory-only is fine */
      }
    }
  }
}

/** Test/introspection helper. */
export function answerCacheStats(): { entries: number; bytes: number } {
  return { entries: mem.size, bytes: memBytes };
}
