// @a0n/.ignored_gnosis/distributed-inference-host/src/audio-stft.ts
// General STFT/iSTFT pair for audio recovery.
// Distinct from mel-spectrogram.ts which is bound to Whisper's params and
// includes mel projection. This module is just plain-magnitude STFT for
// caller-driven frame/hop, with overlap-add inverse and COLA (constant
// overlap-add) verified for the default Hann window. The forward pass
// returns magnitude/phase pairs so callers can apply 2D image-domain
// primitives (deblur, denoise) directly to the time-frequency grid.

export interface STFTConfig {
  /** FFT size — must be a positive power of 2. */
  nFft: number;
  /** Hop length in samples — typically nFft/4 for Hann. */
  hopLength: number;
}

export interface STFTResult {
  /** Magnitude at each (frame, freq). Length = nFrames * (nFft/2 + 1), row-major. */
  magnitude: Float32Array;
  /** Phase at each (frame, freq), radians, in same layout as magnitude. */
  phase: Float32Array;
  nFrames: number;
  nFreqs: number; // nFft / 2 + 1
}

export interface ISTFTArgs {
  magnitude: Float32Array;
  phase: Float32Array;
  nFrames: number;
  config: STFTConfig;
}

function isPowerOfTwo(n: number): boolean {
  return n > 0 && (n & (n - 1)) === 0;
}

const windowCache = new Map<number, Float32Array>();

function hannWindow(n: number): Float32Array {
  const cached = windowCache.get(n);
  if (cached) return cached;
  
  const w = new Float32Array(n);
  if (n === 1) {
    w[0] = 1;
  } else {
    const denom = n - 1;
    for (let i = 0; i < n; i++) {
      w[i] = 0.5 * (1 - Math.cos((2 * Math.PI * i) / denom));
    }
  }
  windowCache.set(n, w);
  return w;
}

/**
 * In-place radix-2 Cooley-Tukey FFT on 1D complex arrays.
 * `re` and `im` must have length === a power of two.
 * `inverse=true` performs the IFFT including the 1/N scaling.
 */
function fft1d(re: Float32Array, im: Float32Array, inverse: boolean): void {
  const n = re.length;
  if (n !== im.length) {
    throw new RangeError("fft1d: re/im length mismatch");
  }
  if (!isPowerOfTwo(n)) {
    throw new RangeError("fft1d: length must be a power of two");
  }
  if (n === 1) {
    return;
  }

  // Bit-reversal permutation.
  let j = 0;
  for (let i = 1; i < n; i++) {
    let bit = n >> 1;
    for (; j & bit; bit >>= 1) {
      j ^= bit;
    }
    j ^= bit;
    if (i < j) {
      const tr = re[i]!;
      re[i] = re[j]!;
      re[j] = tr;
      const ti = im[i]!;
      im[i] = im[j]!;
      im[j] = ti;
    }
  }

  // Butterfly.
  const sign = inverse ? 1 : -1;
  for (let size = 2; size <= n; size <<= 1) {
    const half = size >> 1;
    const tableStep = (sign * 2 * Math.PI) / size;
    for (let i = 0; i < n; i += size) {
      for (let k = 0; k < half; k++) {
        const angle = tableStep * k;
        const wr = Math.cos(angle);
        const wi = Math.sin(angle);
        const a = i + k;
        const b = a + half;
        const xr = re[b]! * wr - im[b]! * wi;
        const xi = re[b]! * wi + im[b]! * wr;
        re[b] = re[a]! - xr;
        im[b] = im[a]! - xi;
        re[a] = re[a]! + xr;
        im[a] = im[a]! + xi;
      }
    }
  }

  if (inverse) {
    const invN = 1 / n;
    for (let i = 0; i < n; i++) {
      re[i] *= invN;
      im[i] *= invN;
    }
  }
}

function validateConfig(config: STFTConfig): void {
  if (!isPowerOfTwo(config.nFft)) {
    throw new RangeError(`stft: nFft must be a power of two, got ${config.nFft}`);
  }
  if (!Number.isFinite(config.hopLength) || config.hopLength <= 0) {
    throw new RangeError(`stft: hopLength must be positive, got ${config.hopLength}`);
  }
  if (config.hopLength > config.nFft) {
    throw new RangeError(
      `stft: hopLength (${config.hopLength}) must not exceed nFft (${config.nFft})`,
    );
  }
}

/**
 * Handles the gnosis stft workflow.
 */
export function stft(audio: Float32Array, config: STFTConfig): STFTResult {
  validateConfig(config);
  const { nFft, hopLength } = config;
  if (audio.length < nFft) {
    throw new RangeError(
      `stft: audio length (${audio.length}) must be >= nFft (${nFft})`,
    );
  }

  // Number of frames such that the last frame ends within the (zero-padded) signal.
  // We zero-pad at the end if needed to fit a whole number of hops.
  const nFrames = Math.floor((audio.length - nFft) / hopLength) + 1;
  const nFreqs = (nFft >> 1) + 1;
  const window = hannWindow(nFft);

  const magnitude = new Float32Array(nFrames * nFreqs);
  const phase = new Float32Array(nFrames * nFreqs);

  const re = new Float32Array(nFft);
  const im = new Float32Array(nFft);

  for (let f = 0; f < nFrames; f++) {
    const start = f * hopLength;
    // Copy windowed frame; zero-pad past audio end if any.
    for (let n = 0; n < nFft; n++) {
      const idx = start + n;
      const sample = idx < audio.length ? audio[idx]! : 0;
      re[n] = sample * window[n]!;
      im[n] = 0;
    }
    fft1d(re, im, false);
    const rowOffset = f * nFreqs;
    for (let k = 0; k < nFreqs; k++) {
      const r = re[k]!;
      const i = im[k]!;
      magnitude[rowOffset + k] = Math.hypot(r, i);
      phase[rowOffset + k] = Math.atan2(i, r);
    }
  }

  return { magnitude, phase, nFrames, nFreqs };
}

/**
 * Returns whether istft is true.
 */
export function istft(args: ISTFTArgs): Float32Array {
  const { magnitude, phase, nFrames, config } = args;
  validateConfig(config);
  const { nFft, hopLength } = config;
  const nFreqs = (nFft >> 1) + 1;

  const expected = nFrames * nFreqs;
  if (magnitude.length !== expected) {
    throw new RangeError(
      `istft: magnitude length ${magnitude.length} does not match nFrames*nFreqs=${expected}`,
    );
  }
  if (phase.length !== expected) {
    throw new RangeError(
      `istft: phase length ${phase.length} does not match nFrames*nFreqs=${expected}`,
    );
  }
  if (nFrames < 1) {
    throw new RangeError(`istft: nFrames must be >= 1, got ${nFrames}`);
  }

  const outputLength = (nFrames - 1) * hopLength + nFft;
  const out = new Float32Array(outputLength);
  const norm = new Float32Array(outputLength);
  const window = hannWindow(nFft);

  const re = new Float32Array(nFft);
  const im = new Float32Array(nFft);

  for (let f = 0; f < nFrames; f++) {
    const rowOffset = f * nFreqs;
    // Build conjugate-symmetric spectrum from magnitude/phase.
    for (let k = 0; k < nFreqs; k++) {
      const m = magnitude[rowOffset + k]!;
      const p = phase[rowOffset + k]!;
      re[k] = m * Math.cos(p);
      im[k] = m * Math.sin(p);
    }
    for (let k = nFreqs; k < nFft; k++) {
      const mirror = nFft - k;
      re[k] = re[mirror]!;
      im[k] = -im[mirror]!;
    }

    fft1d(re, im, true);

    const start = f * hopLength;
    for (let n = 0; n < nFft; n++) {
      const w = window[n]!;
      out[start + n] = out[start + n]! + re[n]! * w;
      norm[start + n] = norm[start + n]! + w * w;
    }
  }

  const eps = 1e-8;
  for (let i = 0; i < outputLength; i++) {
    const d = norm[i]!;
    if (d > eps) {
      out[i] = out[i]! / d;
    }
  }

  return out;
}
