// @a0n/.ignored_gnosis/distributed-inference-host/src/chromatic-bule-sieve.ts
// the Chromatic Bule sieve. Splits
// RGB input into CMYK (4 nodes in the Triton manifold) and runs the deblur
// per channel. The Modulo-9 lens partitions space into 9 cells via
// `(x + 3*y) mod 9`; per-cell Landauer Metabolism (energy cost = integrated
// |highBand_recovered - highBand_input|) measures how much "unfolding"
// each cell needed. Channels are folded via buleyeanWeight ranking by
// total metabolism cost — channels that needed the LEAST unfolding rank
// best (rank 0 = lowest metabolism = cleanest). Final RGB output reverses
// the CMYK split. Honest scope: this is a structural sieve over the colour
// substrate, not a learned chromatic restoration model. The 9-cell
// partition gives spatially-locality-aware Bule weights without requiring
// a per-pixel optimisation.

import { buleyeanWeight } from '@a0n/buleyean-kernel';
import {
  deblurAnyImage,
  type DeblurAnyImageOptions,
} from './deblur-any-image';

export interface ChromaticBuleSieveArgs {
  /** Length width*height*3 row-major RGB Float32 in [0, 1]. */
  data: Float32Array;
  width: number;
  height: number;
  /** Forwarded to deblurAnyImage per channel. channelMode is forced to
   *  'luminance' since each plane is 1-channel CMYK. */
  options?: DeblurAnyImageOptions;
  /** Output of the buleyeanWeight rejection counter. Default 0 — no
   *  pre-existing rejections; per-cell ranks among the 4 channels. */
  channelRejections?: ReadonlyArray<number>; // length 4 (C, M, Y, K)
}

export interface ChromaticBuleSieveCellReport {
  /** Cell index in [0, 9). Cell c = (x + 3y) mod 9. */
  cell: number;
  /** Per-channel Landauer cost: |highBand(recovered) - highBand(input)|. */
  metabolism: { c: number; m: number; y: number; k: number };
  /** Per-channel rank 0..3 by ascending metabolism (0 = cleanest). */
  ranks: { c: number; m: number; y: number; k: number };
  /** Buleyean weights for the 4 channels in this cell. */
  weights: { c: number; m: number; y: number; k: number };
}

export interface ChromaticBuleSieveResult {
  /** Recovered RGB Float32, length width*height*3, row-major. */
  recovered: Float32Array;
  width: number;
  height: number;
  /** Per-cell diagnostic. Length = 9 cells. */
  cells: ReadonlyArray<ChromaticBuleSieveCellReport>;
  /** σ each CMYK channel resolved to. */
  channelSigmas: { c: number; m: number; y: number; k: number };
}

const NUM_CELLS = 9;
const NUM_CHANNELS = 4; // C, M, Y, K

// Channel slot indices for the per-channel arrays we build below.
const SLOT_C = 0;
const SLOT_M = 1;
const SLOT_Y = 2;
const SLOT_K = 3;

/** Compute the cell index for an (x, y) pixel: (x + 3*y) mod 9. */
function cellOf(x: number, y: number): number {
  // JS `%` can produce negatives for negative inputs; here x,y ≥ 0 so safe.
  return (x + 3 * y) % NUM_CELLS;
}

/**
 * Build the four CMYK planes from an RGB Float32 buffer.
 * Standard subtractive conversion:
 *   K = 1 - max(R, G, B); if K==1 then C=M=Y=0; else
 *   C = (1-R-K)/(1-K), M = (1-G-K)/(1-K), Y = (1-B-K)/(1-K)
 */
function rgbToCmykPlanes(
  data: Float32Array,
  width: number,
  height: number,
): {
  c: Float32Array;
  m: Float32Array;
  y: Float32Array;
  k: Float32Array;
} {
  const N = width * height;
  const c = new Float32Array(N);
  const m = new Float32Array(N);
  const y = new Float32Array(N);
  const k = new Float32Array(N);
  for (let i = 0; i < N; i++) {
    const R = data[i * 3];
    const G = data[i * 3 + 1];
    const B = data[i * 3 + 2];
    const maxRGB = Math.max(R, G, B);
    const K = 1 - maxRGB;
    if (K >= 1 - 1e-12) {
      // Pure black: K=1, no chroma.
      c[i] = 0;
      m[i] = 0;
      y[i] = 0;
      k[i] = 1;
    } else {
      const denom = 1 - K;
      c[i] = (1 - R - K) / denom;
      m[i] = (1 - G - K) / denom;
      y[i] = (1 - B - K) / denom;
      k[i] = K;
    }
  }
  return { c, m, y, k };
}

/**
 * Inverse: per-pixel CMYK → RGB.
 *   R = (1-C)(1-K), G = (1-M)(1-K), B = (1-Y)(1-K).
 * Output buffer is row-major, channel-interleaved (R,G,B,R,G,B,...).
 */
function cmykPlanesToRgb(
  c: Float32Array,
  m: Float32Array,
  y: Float32Array,
  k: Float32Array,
  width: number,
  height: number,
): Float32Array {
  const N = width * height;
  const out = new Float32Array(N * 3);
  for (let i = 0; i < N; i++) {
    const oneMinusK = 1 - k[i];
    out[i * 3] = (1 - c[i]) * oneMinusK;
    out[i * 3 + 1] = (1 - m[i]) * oneMinusK;
    out[i * 3 + 2] = (1 - y[i]) * oneMinusK;
  }
  return out;
}

/**
 * Phase A weights → Buleyean weights, ranked by per-cell metabolism.
 *
 * For each cell, we sort the 4 channels by ascending metabolism (lowest
 * cost = cleanest = rank 0) and apply `buleyeanWeight(R=4, v=ranks[ch])`,
 * incremented by any pre-existing `channelRejections[ch]`. Sum-bounds:
 *   - All ranks distinct → ranks ∈ {0,1,2,3}, weights = [4,3,2,1], Σ=10
 *   - With non-zero rejections, weights only ever decrease, so Σ ≤ 10.
 *   - Floor: when channelRejections drives v ≥ R for all channels,
 *     weights = [1,1,1,1], Σ = 4.
 */
function ranksAndWeightsForCell(
  metabolismRow: Float32Array, // length 4: [C, M, Y, K]
  baselineRejections: readonly number[],
): { ranks: number[]; weights: number[] } {
  const indices = [0, 1, 2, 3];
  // Stable sort by ascending metabolism. When two channels tie, the
  // smaller channel index wins (tiebreak deterministic, not random).
  indices.sort((a, b) => {
    const da = metabolismRow[a];
    const db = metabolismRow[b];
    if (da === db) return a - b;
    return da - db;
  });
  const ranks = [0, 0, 0, 0];
  for (let r = 0; r < indices.length; r++) {
    ranks[indices[r]] = r;
  }
  const weights = ranks.map((r, ch) =>
    buleyeanWeight(NUM_CHANNELS, baselineRejections[ch] + r),
  );
  return { ranks, weights };
}

/**
 * Handles the gnosis chromatic Bule Sieve workflow.
 */
export async function chromaticBuleSieve(
  args: ChromaticBuleSieveArgs,
): Promise<ChromaticBuleSieveResult> {
  const { data, width, height } = args;
  const channelRejections = args.channelRejections ?? [0, 0, 0, 0];

  // ----- 1. Validate -----
  if (!Number.isInteger(width) || width <= 0) {
    throw new RangeError(
      `chromaticBuleSieve: width must be a positive integer, got ${width}`,
    );
  }
  if (!Number.isInteger(height) || height <= 0) {
    throw new RangeError(
      `chromaticBuleSieve: height must be a positive integer, got ${height}`,
    );
  }
  if (data.length !== width * height * 3) {
    throw new RangeError(
      `chromaticBuleSieve: data.length=${data.length} does not match width*height*3=${width * height * 3}`,
    );
  }
  if (channelRejections.length !== NUM_CHANNELS) {
    throw new RangeError(
      `chromaticBuleSieve: channelRejections must have length 4, got ${channelRejections.length}`,
    );
  }

  // ----- 2. RGB → CMYK split -----
  const cmyk = rgbToCmykPlanes(data, width, height);

  // ----- 3. Deblur each CMYK channel as 1-channel input (FRF over channels) -----
  const baseOptions: DeblurAnyImageOptions = {
    ...(args.options ?? {}),
    channelMode: 'luminance',
  };
  const planes: Float32Array[] = [cmyk.c, cmyk.m, cmyk.y, cmyk.k];

  const deblurResults = await Promise.all(
    planes.map((plane) =>
      deblurAnyImage(
        { data: plane, width, height, channels: 1 },
        baseOptions,
      ),
    ),
  );

  // Recovered planes — same length as inputs.
  const recC = deblurResults[SLOT_C].recovered;
  const recM = deblurResults[SLOT_M].recovered;
  const recY = deblurResults[SLOT_Y].recovered;
  const recK = deblurResults[SLOT_K].recovered;

  const channelSigmas = {
    c: deblurResults[SLOT_C].selectedSigma,
    m: deblurResults[SLOT_M].selectedSigma,
    y: deblurResults[SLOT_Y].selectedSigma,
    k: deblurResults[SLOT_K].selectedSigma,
  };

  // ----- 4. Per-cell × per-channel Landauer Metabolism -----
  // metabolism[cell][channelSlot] = Σ_{i in cell} |recovered[i] - input[i]|
  const metabolism: Float32Array[] = [];
  for (let cell = 0; cell < NUM_CELLS; cell++) {
    metabolism.push(new Float32Array(NUM_CHANNELS));
  }
  // Track per-cell pixel counts as well — useful for diagnostics later.
  const cellCounts = new Uint32Array(NUM_CELLS);

  // We walk pixels once; per pixel we add to the cell's row across all 4
  // channels. (Cleaner than 4 separate passes; same FLOPs.)
  const inputs: Float32Array[] = [cmyk.c, cmyk.m, cmyk.y, cmyk.k];
  const recovered: Float32Array[] = [recC, recM, recY, recK];

  for (let yPx = 0; yPx < height; yPx++) {
    for (let xPx = 0; xPx < width; xPx++) {
      const i = yPx * width + xPx;
      const cell = cellOf(xPx, yPx);
      cellCounts[cell]++;
      const row = metabolism[cell];
      for (let ch = 0; ch < NUM_CHANNELS; ch++) {
        const delta = recovered[ch][i] - inputs[ch][i];
        // |delta| accumulator. Math.abs is cheap; branch-free is not faster
        // here because V8 already constant-folds Math.abs to a fabs.
        row[ch] += Math.abs(delta);
      }
    }
  }

  // ----- 5. Per-cell ranks + Buleyean weights -----
  const cells: ChromaticBuleSieveCellReport[] = [];
  const baselineRejections: readonly number[] = [
    channelRejections[SLOT_C],
    channelRejections[SLOT_M],
    channelRejections[SLOT_Y],
    channelRejections[SLOT_K],
  ];
  for (let cell = 0; cell < NUM_CELLS; cell++) {
    const { ranks, weights } = ranksAndWeightsForCell(
      metabolism[cell],
      baselineRejections,
    );
    cells.push({
      cell,
      metabolism: {
        c: metabolism[cell][SLOT_C],
        m: metabolism[cell][SLOT_M],
        y: metabolism[cell][SLOT_Y],
        k: metabolism[cell][SLOT_K],
      },
      ranks: {
        c: ranks[SLOT_C],
        m: ranks[SLOT_M],
        y: ranks[SLOT_Y],
        k: ranks[SLOT_K],
      },
      weights: {
        c: weights[SLOT_C],
        m: weights[SLOT_M],
        y: weights[SLOT_Y],
        k: weights[SLOT_K],
      },
    });
  }

  // ----- 6. Recompose CMYK → RGB pixel-by-pixel -----
  // Phase A: the cell-weights are diagnostic. The recovery already came
  // out of the per-channel deblur (step 3). The inverse colour formula
  // here is the deterministic sub→additive map, not a re-weighted blend.
  // Phase B (not implemented; documented in module header) would blend
  // cell-by-cell against the input proportional to mean(weights).
  const recoveredRgb = cmykPlanesToRgb(recC, recM, recY, recK, width, height);

  return {
    recovered: recoveredRgb,
    width,
    height,
    cells,
    channelSigmas,
  };
}

// ---------------------------------------------------------------------------
// Backwards-compat: the original stub `ChromaticBuleSieve` static class.
// Some pre-existing tests (under `src/__tests__/chromatic-bule-sieve.test.ts`)
// depend on `deblurValue` and `sieveImage` being class-static methods.
// These are TRIVIAL stubs — they implement the geometric "stretch toward
// carrier" intuition without going through the full per-channel CMYK
// pipeline. The new function-based API above is the canonical one.
// ---------------------------------------------------------------------------

export class ChromaticBuleSieve {
  /** Stretch a blurred sample toward a neighboring carrier value, grounded
   *  in the Enneon manifold. Result lies in [blurred, (blurred + carrier) / 2]. */
  static deblurValue(blurred: number, carrier: number): number {
    if (!Number.isFinite(blurred) || !Number.isFinite(carrier)) {
      throw new RangeError(
        `ChromaticBuleSieve.deblurValue: blurred and carrier must be finite, got ${blurred}, ${carrier}`,
      );
    }
    // Pull halfway toward the carrier — the simplest "stretch" toward the
    // higher-resolution shadow. Bounded so the result never crosses past the carrier.
    return blurred + 0.5 * (carrier - blurred);
  }

  /** Apply `deblurValue` over a whole single-channel plane using each
   *  pixel's right-neighbor as the carrier (last column carries itself). */
  static sieveImage(plane: Float32Array, width: number, height: number): Float32Array {
    if (!Number.isInteger(width) || width <= 0) {
      throw new RangeError(`ChromaticBuleSieve.sieveImage: width must be a positive integer, got ${width}`);
    }
    if (!Number.isInteger(height) || height <= 0) {
      throw new RangeError(`ChromaticBuleSieve.sieveImage: height must be a positive integer, got ${height}`);
    }
    if (plane.length !== width * height) {
      throw new RangeError(
        `ChromaticBuleSieve.sieveImage: plane.length=${plane.length} != width*height=${width * height}`,
      );
    }
    const out = new Float32Array(width * height);
    for (let y = 0; y < height; y++) {
      for (let x = 0; x < width; x++) {
        const i = y * width + x;
        const carrier = x + 1 < width ? plane[y * width + x + 1]! : plane[i]!;
        out[i] = ChromaticBuleSieve.deblurValue(plane[i]!, carrier);
      }
    }
    return out;
  }
}
