// @a0n/.ignored_gnosis/distributed-inference-host/src/codec-ancestry.ts
/**
 * # Codec Ancestry — lossy re-encoding leaves a spectral cutoff fingerprint
 *
 * Mirrors `Gnosis.Mesh.CodecAncestry` from `open-source/gnosis-math`, the Rustic
 * Church carrier for codec-ancestry detection, framed *defensively* as a
 * transcode / re-upload detector. Lossy perceptual codecs (MP3, AAC, Opus) do
 * not preserve a signal's full spectrum: to save bits they apply a hard low-pass
 * — high-frequency bands above some cutoff are forced to zero well below the
 * Nyquist limit — and they quantize, punching holes. A recording that was once
 * compressed therefore carries a structural fingerprint of its ancestry: an
 * all-zero high-frequency tail sitting above bands that still carry energy. A
 * truly *pristine* original recorded at full bandwidth has no such tail.
 *
 * A "this audio is a pristine original" provenance claim is exactly the thing a
 * re-uploader / launderer wants to forge, so the Skeptic admits the pristine
 * claim only when (1) the spectrum actually carries energy — silence is not
 * "pristine audio", it is nothing — AND (2) no compression-ancestry artifact (a
 * high-band cutoff) is present. A transcode tightens the cutoff (pushes it to a
 * lower band), and re-encoding never restores the lost high band, so a flagged
 * spectrum stays flagged under further encoding.
 *
 * ## What is (and is not) carried
 *
 * This module is the runtime promotion check for a codec-ancestry observation.
 * It carries ONLY the finite contract the Lean file proves:
 *
 *  - the per-band spectrum with its summed energy (`totalEnergy`);
 *  - the support gate (the source-only rejection analog): an all-zero spectrum
 *    carries no audio and is withheld (`hasSupport`);
 *  - the cutoff gate (the ancestry artifact): a compression cutoff is present
 *    when, after some lower band carried energy, the remaining high-frequency
 *    suffix is nonempty and entirely zero (`hasCutoff`);
 *  - pristine admission: a pristine claim is admitted only when the spectrum has
 *    support AND no cutoff (`admitPristine`).
 *
 * Real STFT band estimation, psychoacoustic cutoff localization, quantization-
 * hole statistics, joint-stereo and SBR artifacts, and the actual codec-
 * population separation stay runtime/empirical obligations OUTSIDE this
 * certificate — they are not theorems and are not asserted here.
 *
 * The promotion check folds the contract into the same epistemic-outcome
 * vocabulary the sibling mesh carriers already speak
 * (`claim_admitted` / `observation_only` / `claim_withheld`), so a recording
 * cannot be displayed as a pristine-original claim until it has cleared the
 * Lean-side support/cutoff gate.
 *
 * ## Lean source-of-truth
 *
 * @see open-source/gnosis-math/Gnosis/Mesh/CodecAncestry.lean
 * @see open-source/gnosis-math/Gnosis/Mesh/ElectricalNetworkFrequency.lean (List sibling)
 */

/* ------------------------------------------------------------------ */
/*  Spectra and the support gate                                      */
/* ------------------------------------------------------------------ */

/**
 * A spectrum is a list of per-band magnitudes over a fixed window; index ↑ means
 * frequency ↑, so the *tail* of the list holds the high-frequency bands.
 * Magnitudes are non-negative integers (saturating `Nat` arithmetic), matching
 * the Lean `Spectrum`.
 */
export type Spectrum = readonly number[];

/**
 * Total energy of a spectrum: the summed per-band magnitudes. An all-zero
 * spectrum has zero total energy and carries no audio.
 * Mirrors `CodecAncestry.totalEnergy`.
 */
export function totalEnergy(s: Spectrum): number {
  let energy = 0;
  for (const band of s) {
    energy += band;
  }
  return energy;
}

/**
 * A spectrum has support when it carries some energy (some band is nonzero).
 * Mirrors `CodecAncestry.HasSupport` / `hasSupportB`.
 */
export function hasSupport(s: Spectrum): boolean {
  return totalEnergy(s) > 0;
}

/* ------------------------------------------------------------------ */
/*  Cutoff gate (the ancestry artifact)                               */
/* ------------------------------------------------------------------ */

/** Every band of `s` is zero. Mirrors `CodecAncestry.allZeroB`. */
function allZero(s: Spectrum): boolean {
  for (const band of s) {
    if (band !== 0) {
      return false;
    }
  }
  return true;
}

/**
 * `hasCutoff` scans from the low-frequency end carrying a flag recording whether
 * any lower band has already shown energy. A compression cutoff is present
 * exactly when, *after* some lower band carried energy, the remaining
 * high-frequency suffix is nonempty and entirely zero — the silenced tail
 * sitting above the surviving low bands. A full-band spectrum (top band nonzero)
 * never reaches an all-zero nonempty suffix preceded by energy.
 *
 * Mirrors `CodecAncestry.hasCutoffB` (top-level scan from `energyBelow = false`).
 */
export function hasCutoff(s: Spectrum): boolean {
  let energyBelow = false;
  for (let i = 0; i < s.length; i++) {
    const suffix = s.slice(i);
    // A cutoff is found here when energy was seen below AND the whole remaining
    // suffix is a nonempty all-zero tail.
    if (energyBelow && suffix.length > 0 && allZero(suffix)) {
      return true;
    }
    if (s[i] > 0) {
      energyBelow = true;
    }
  }
  return false;
}

/**
 * A spectrum showing a cutoff is flagged as a transcode (the ancestry artifact
 * present — the negation of pristine cutoff-free).
 * Mirrors `CodecAncestry.transcodeFlaggedB`.
 */
export function transcodeFlagged(s: Spectrum): boolean {
  return hasCutoff(s);
}

/* ------------------------------------------------------------------ */
/*  Runtime promotion check (epistemic outcome)                       */
/* ------------------------------------------------------------------ */

/**
 * Epistemic outcome of a codec-ancestry observation. Identical vocabulary to the
 * sibling mesh carriers (`./electrical-network-frequency`,
 * `./cross-node-coincidence`), kept as a local literal union so this certificate
 * has no import coupling:
 *
 *  - `claim_admitted`    — the spectrum carries energy AND shows no compression-
 *                          ancestry cutoff, so the pristine-original claim may be
 *                          promoted.
 *  - `observation_only`  — reserved for completeness; codec ancestry has no
 *                          intermediate "reported but uncertified" state (a
 *                          cutoff is a hard ancestry artifact, not a soft
 *                          suspicion), so it is not emitted by `admitPristine`.
 *  - `claim_withheld`    — the spectrum is silent (no support, nothing to call an
 *                          original) or carries a compression cutoff (a transcode
 *                          / re-upload, which must be re-uploaded or transcoded
 *                          from a true original to be certified).
 */
export type CodecAncestryEpistemicOutcome =
  | 'claim_admitted'
  | 'observation_only'
  | 'claim_withheld';

/** The reason a pristine-original claim was held back from promotion. */
export type CodecAncestryLimitationClass =
  | 'no_audio'
  | 'compression_ancestry'
  | 'none';

export interface CodecAncestryAdmission {
  readonly outcome: CodecAncestryEpistemicOutcome;
  readonly limitation: CodecAncestryLimitationClass;
  /** Whether the spectrum carries any energy (passed the support gate). */
  readonly supported: boolean;
  /** Whether a compression-ancestry cutoff was detected. */
  readonly cutoff: boolean;
  /** The measured total energy (the displayable loudness magnitude). */
  readonly totalEnergy: number;
  /** Operator-facing remediation when the claim is withheld for ancestry. */
  readonly remediation?: string;
  readonly summary: string;
}

/**
 * The promotion check Monster Mesh runs before any UI surface displays a
 * recording as a pristine original. It folds the Lean contract into an epistemic
 * outcome:
 *
 *  - the support gate decides whether there is any audio at all — an all-zero
 *    spectrum is silence and is withheld (the source-only rejection analog,
 *    mirroring `zeroSpec_not_admitted`);
 *  - the cutoff gate decides whether a compression-ancestry artifact is present —
 *    a silenced high tail above surviving energy is withheld as a transcode /
 *    re-upload (mirroring `encoded_not_admitted`), with the remediation being to
 *    re-upload or transcode from a true original;
 *  - a supported, cutoff-free spectrum is admitted as pristine (mirroring
 *    `pristine_admitted`).
 *
 * @param spectrum the per-band magnitude spectrum under examination
 */
export function admitPristine(spectrum: Spectrum): CodecAncestryAdmission {
  const energy = totalEnergy(spectrum);
  const supported = energy > 0;
  const cutoff = hasCutoff(spectrum);

  // Support gate: an all-zero spectrum is silence — there is no original to
  // certify, exactly as a silent ENF trace carries no timestamp.
  if (!supported) {
    return {
      outcome: 'claim_withheld',
      limitation: 'no_audio',
      supported,
      cutoff,
      totalEnergy: energy,
      summary:
        'all-zero spectrum: no energy carries no audio — there is no original to certify',
    };
  }

  // Cutoff gate: a silenced high tail above surviving energy is a compression-
  // ancestry artifact, so the pristine claim is withheld; the path back to a
  // claim is a re-upload / transcode from a true original.
  if (cutoff) {
    return {
      outcome: 'claim_withheld',
      limitation: 'compression_ancestry',
      supported,
      cutoff,
      totalEnergy: energy,
      remediation:
        're-upload or transcode from a true full-bandwidth original — re-encoding never restores the lost high band',
      summary:
        'compression-ancestry cutoff detected: an all-zero high-frequency tail sits above surviving energy — this is a transcode / re-upload, not a pristine original',
    };
  }

  return {
    outcome: 'claim_admitted',
    limitation: 'none',
    supported,
    cutoff,
    totalEnergy: energy,
    summary: `full-band spectrum (total energy ${energy}) carries energy with no high-band cutoff — pristine original certified`,
  };
}

/* ------------------------------------------------------------------ */
/*  Concrete fixtures (mirror the Lean fixtures)                      */
/* ------------------------------------------------------------------ */

/** A pristine, full-band original: every band carries energy, the top band
 *  included, so there is no silenced high tail. Total energy `5+7+6+4+3+2 = 27`.
 *  Mirrors `CodecAncestry.pristine`. */
export const pristine: Spectrum = [5, 7, 6, 4, 3, 2];

/** The SAME underlying content after one lossy encode: the two top bands have
 *  been forced to zero (the codec's hard low-pass), while the low bands `5,7,6,4`
 *  survive. Total energy `22 > 0` (support holds), but the nonempty all-zero high
 *  tail `[0,0]` sits above energy → a cutoff is present.
 *  Mirrors `CodecAncestry.encoded`. */
export const encoded: Spectrum = [5, 7, 6, 4, 0, 0];

/** A transcode of `encoded` at a more aggressive bitrate: the cutoff tightens,
 *  zeroing the top FOUR bands, leaving only `5,7`. Still has support (`12`), and
 *  an even larger all-zero tail `[0,0,0,0]` above energy → still flagged.
 *  Mirrors `CodecAncestry.transcode`. */
export const transcode: Spectrum = [5, 7, 0, 0, 0, 0];

/** An all-zero spectrum: silence. No energy, so no pristine claim either way —
 *  this is nothing, not a pristine original. Mirrors `CodecAncestry.zeroSpec`. */
export const zeroSpec: Spectrum = [0, 0, 0, 0, 0, 0];

/**
 * Codec-Ancestry master witness. Bundles the boundary facts the Lean theorem
 * `codec_ancestry_master` proves, each recomputed here from first principles
 * (never imported from Lean IR).
 *
 * Mirrors `CodecAncestry.codec_ancestry_master`.
 */
export const CodecAncestryMaster = {
  /** The pristine original carries energy: it has support. */
  pristineHasSupport: hasSupport(pristine) === true,

  /** The all-zero spectrum is rejected by the support gate. */
  zeroSpecNoSupport: hasSupport(zeroSpec) === false,

  /** The pristine full-band spectrum shows NO cutoff. */
  pristineNoCutoff: hasCutoff(pristine) === false,

  /** The encoded spectrum shows a cutoff (energy then all-zero tail `[0,0]`). */
  encodedHasCutoff: hasCutoff(encoded) === true,

  /** The transcode shows a cutoff too (a tighter all-zero tail `[0,0,0,0]`). */
  transcodeHasCutoff: hasCutoff(transcode) === true,

  /** The pristine original is admitted: support AND no cutoff. */
  pristineAdmitted: admitPristine(pristine).outcome === 'claim_admitted',

  /** The encoded copy is NOT admitted as pristine: it fails the cutoff gate. */
  encodedNotAdmitted: admitPristine(encoded).outcome !== 'claim_admitted',

  /** The all-zero spectrum is NOT admitted as pristine: it fails the support gate. */
  zeroSpecNotAdmitted: admitPristine(zeroSpec).outcome !== 'claim_admitted',

  /** The encoded copy is flagged as a transcode (ancestry artifact present). */
  encodedFlagged: transcodeFlagged(encoded) === true,

  /** The transcode is flagged as well: a tighter cutoff is still a cutoff. */
  transcodeFlagged: transcodeFlagged(transcode) === true,

  /** A pristine original is NOT flagged: no ancestry artifact. */
  pristineNotFlagged: transcodeFlagged(pristine) === false,

  /** Hard negative: pristine and encoded share their first four (low) bands, so
   *  the distinction is the ancestry artifact, not the content. */
  fixturesShareLowBands:
    JSON.stringify(pristine.slice(0, 4)) ===
    JSON.stringify(encoded.slice(0, 4)),
} as const;
