// @a0n/.ignored_gnosis/distributed-inference-host/src/conflict-register.ts
/**
 * conflict-register.ts
 * ====================
 *
 * TS host port of the twin-forge conflict-register steerer
 * (`apps/pensieve-ios-importer/twin-forge/conflict-steering.ts`). The orthographic
 * scorer (`personality-orthography.ts`) steers HOW a contact writes (caps/punct/
 * emoji); this steers WHAT register they reach for under conflict — the SIGNED
 * semantic bias from the abductive spectral model.
 *
 * A register word loads onto the four wave-channels with a signed affinity. The
 * contact's signed channel gains × that affinity, clamped into ±cap (signed, no
 * −∞ — anti-tells DEMOTE but never forbid, per `Gnosis.PersonalityLogitSteering`),
 * give the per-token bias. We resolve the lexicon WORDS to the model's token ids
 * over its own vocab (`vocab[id]` = surface string), biasing only WORD-INITIAL
 * pieces — biasing a bare sub-word fragment ("fine" inside "define") would corrupt
 * unrelated words. The result is a dense `number[]` aligned to the vocab so it
 * SUMS elementwise into the orthographic `boundary_bias` vector the fat-station
 * applies before the argmax (`model.rs::decode_next_fast_argmax_once`).
 *
 * Wire shape (`PersonalitySteeringSpec.semantic`, from the edge upload route's
 * `normalizeSemanticGains`): { behavioral, aesthetic, prosody, harmonic, cap } —
 * the spectral `SignedGains`. Kept identical to twin-forge so both ends agree.
 */

export interface SemanticSteeringGains {
  behavioral: number;
  aesthetic: number;
  prosody: number;
  harmonic: number;
  /** signed-bias envelope cap (dose–response; ≈ ½ the logit margin to overturn). */
  cap: number;
}

type Aff = { behavioral?: number; aesthetic?: number; prosody?: number; harmonic?: number };

/** Conflict-register lexicon — kept in lockstep with twin-forge conflict-steering.ts. */
const LEXICON: Array<{ tokens: string[]; aff: Aff }> = [
  // assertive / dominant / competing  → behavioral +
  { tokens: ['no', 'not', 'never', 'stop', "won't", "can't", "don't", 'enough', 'done', 'absolutely', 'period', 'refuse'], aff: { behavioral: 1.0 } },
  // contempt / dismissive  → behavioral +, aesthetic −
  { tokens: ['seriously', 'obviously', 'clearly', 'typical', 'unbelievable', 'pathetic', 'ridiculous', 'whatever'], aff: { behavioral: 0.8, aesthetic: -0.5 } },
  // cold / withholding / curt  → behavioral −, aesthetic −  (the avoiding voice)
  { tokens: ['fine', 'k', 'ok', 'sure', 'busy', 'later', 'cool', 'noted', "we'll see"], aff: { behavioral: -0.7, aesthetic: -0.5 } },
  // warm / affectionate  → aesthetic +, behavioral −
  { tokens: ['love', 'miss', 'thank', 'thanks', 'appreciate', 'sweet', 'proud', 'care', 'hug', '❤', '🥰', '😘', '🙏'], aff: { aesthetic: 1.0, behavioral: -0.3 } },
  // repair / accommodating  → aesthetic + (WARMTH is the separating axis), behavioral − slight.
  { tokens: ['sorry', 'apologize', 'my bad', "you're right", 'understand', 'fair', 'forgive', "i hear you"], aff: { behavioral: -0.3, aesthetic: 0.8 } },
  // anxious / reassurance-seeking  → prosody −
  { tokens: ['why', 'please', 'still', 'are you', 'okay?', 'ok?', 'right?', 'are we'], aff: { prosody: -0.9 } },
  // open / exploratory / collaborating  → harmonic +
  { tokens: ['how about', 'what if', "let's", 'maybe we', 'together', 'idea', 'thought'], aff: { harmonic: 0.8, behavioral: 0.2 } },
];

const tokenAff = new Map<string, Aff>();
for (const { tokens, aff } of LEXICON) for (const t of tokens) tokenAff.set(t, aff);

const clampSym = (x: number, cap: number) => Math.max(-cap, Math.min(cap, x));

/** Per-word signed bias from the gains (0 = identity, no register opinion). */
function wordBias(word: string, g: SemanticSteeringGains): number {
  const t = word.toLowerCase().trim();
  const aff = tokenAff.get(t) ?? [...tokenAff.entries()].find(([k]) => k.includes(' ') && t.startsWith(k.split(' ')[0]!))?.[1];
  if (!aff) return 0;
  const raw =
    g.behavioral * (aff.behavioral ?? 0) +
    g.aesthetic * (aff.aesthetic ?? 0) +
    g.prosody * (aff.prosody ?? 0) +
    g.harmonic * (aff.harmonic ?? 0);
  return clampSym(Math.round(raw * g.cap), g.cap);
}

// Normalize a vocab piece to its word form. The HOST tokenizer's `decodeToken`
// already replaces the metaspace/word-initial marker ('Ġ' U+0120 byte-level, '▁'
// U+2581 SentencePiece) with a space and STRIPS the leading space — so live vocab
// strings arrive as clean decoded words with NO marker ("love", not "▁love"). We
// therefore match the whole decoded token; the leading-marker strip below is a
// belt-and-suspenders for any tokenizer that DOES preserve the marker. (This is
// exactly how the working orthographic compiler treats the same strings.)
const stripMarker = (piece: string): string => piece.replace(/^[Ġ▁\s]+/, '').toLowerCase();

/**
 * Compile a dense per-token semantic bias vector (length = vocab.length) from the
 * signed gains and the model's vocab strings. A token whose decoded word matches a
 * conflict-register lexicon entry is biased; everything else is 0. Because the host
 * strips the word-initial marker, we cannot distinguish a word-initial token from a
 * (rare) bare sub-word fragment that decodes to the same letters — we accept that
 * low-harm false positive in exchange for the bias actually landing. Designed to be
 * summed elementwise into the orthographic `boundary_bias` (Path-A install vector).
 */
export function compileSemanticBias(
  gains: SemanticSteeringGains,
  vocab: readonly string[],
): number[] {
  const out = new Array<number>(vocab.length).fill(0);
  if (!Number.isFinite(gains.cap) || gains.cap <= 0) return out;
  for (let id = 0; id < vocab.length; id++) {
    const piece = vocab[id]!;
    if (!piece) continue;
    const word = stripMarker(piece);
    if (!word) continue;
    out[id] = wordBias(word, gains);
  }
  return out;
}

/** Sum two equal-length bias vectors elementwise (orthographic + semantic), into a
 * single Path-A vector. Lengths default to the longer; missing entries are 0. */
export function combineBias(a: readonly number[], b: readonly number[]): number[] {
  const n = Math.max(a.length, b.length);
  const out = new Array<number>(n).fill(0);
  for (let i = 0; i < n; i++) out[i] = (a[i] ?? 0) + (b[i] ?? 0);
  return out;
}
