// @a0n/.ignored_gnosis/distributed-inference-host/src/cost-algebra.ts
/**
 * # The Algebra of Cost — runtime core
 *
 * Mirrors `Gnosis.CostAlgebra` from `open-source/gnosis-math`. A
 * `CostAlgebra<S>` is a commutative monoid `(S, ·, e)` together with a
 * monoid homomorphism `score : S → (Number, +, 0)`. The Bule unit
 * (`BuleyProjection`) is the canonical instance via componentwise add.
 *
 * Conservation is *baked into the morphism condition*: a `CostHom` must
 * preserve zero, composition, and score. This makes information
 * conservation a categorical fact, not a runtime convention.
 *
 * The runtime mirror does not re-prove the laws — those live in the
 * Lean modules cited below — but it exposes the same primitives and
 * (where helpful) cross-checks via tests.
 *
 * @see open-source/gnosis-math/Gnosis/CostAlgebra.lean
 * @see open-source/gnosis-math/Gnosis/CostAlgebraDerivations.lean
 * @see open-source/gnosis-math/Gnosis/CostAlgebraCategory.lean
 * @see open-source/gnosis-math/Gnosis/CostAlgebraTerminalInitial.lean
 */

import type { BuleyFace, BuleyProjection } from './spectral-noise-monitor';
import { vacuumBuleyProjection } from './spectral-noise-monitor';

/** A commutative monoid (zero, compose) with a score morphism into Nat. */
export interface CostAlgebra<S> {
  zero: S;
  compose(a: S, b: S): S;
  score(s: S): number;
  equals(a: S, b: S): boolean;
}

/** A score-preserving homomorphism between cost algebras. */
export interface CostHom<S, T> {
  map(s: S): T;
  /** Identity-witness: source's cost-algebra (referential, for type pinning). */
  source: CostAlgebra<S>;
  target: CostAlgebra<T>;
}

/** The trivial 1-faced CostAlgebra: `(Number, +, 0)` with score = id. */
export const natCostAlgebra: CostAlgebra<number> = {
  zero: 0,
  compose: (a, b) => a + b,
  score: (n) => n,
  equals: (a, b) => a === b,
};

/** Componentwise add on `BuleyProjection`. */
function addBuleyProjections(a: BuleyProjection, b: BuleyProjection): BuleyProjection {
  const waste = a.waste + b.waste;
  const opportunity = a.opportunity + b.opportunity;
  const diversity = a.diversity + b.diversity;
  return { waste, opportunity, diversity, score: waste + opportunity + diversity };
}

/** Equality for `BuleyProjection` — face-by-face. The `score` field is
 *  derived, but compared explicitly to catch desync. */
function equalsBuleyProjections(a: BuleyProjection, b: BuleyProjection): boolean {
  return (
    a.waste === b.waste &&
    a.opportunity === b.opportunity &&
    a.diversity === b.diversity &&
    a.score === b.score
  );
}

/** The Bule unit cost algebra. Mirrors `Gnosis.CostAlgebra.buleyCostAlgebra`. */
export const buleyCostAlgebra: CostAlgebra<BuleyProjection> = {
  zero: vacuumBuleyProjection,
  compose: addBuleyProjections,
  score: (b) => b.score,
  equals: equalsBuleyProjections,
};

/** The canonical hom into `natCostAlgebra` — the score function itself.
 *  Mirrors `Gnosis.CostAlgebraCategory.scoreHom`. By
 *  `nat_terminal_unique`, this is the *only* hom into `natCostAlgebra`,
 *  making `natCostAlgebra` the terminal object of the cost-algebra
 *  category. */
export function scoreHom<S>(A: CostAlgebra<S>): CostHom<S, number> {
  return {
    map: (s) => A.score(s),
    source: A,
    target: natCostAlgebra,
  };
}

/** Identity hom on a cost algebra. Mirrors `Gnosis.CostAlgebraCategory.idHom`. */
export function idHom<S>(A: CostAlgebra<S>): CostHom<S, S> {
  return { map: (s) => s, source: A, target: A };
}

/** Composition of homs. Mirrors `Gnosis.CostAlgebraCategory.composeHom`. */
export function composeHom<S, T, U>(
  f: CostHom<S, T>,
  g: CostHom<T, U>,
): CostHom<S, U> {
  return {
    map: (s) => g.map(f.map(s)),
    source: f.source,
    target: g.target,
  };
}

/** A `BuleyFace`-indexed unit element with score 1 — the per-face
 *  clinamen lift in monoid form. Mirrors
 *  `Gnosis.CostAlgebra.buleyFaceUnit`. */
export function buleyFaceUnit(face: BuleyFace): BuleyProjection {
  const proto = { ...vacuumBuleyProjection };
  proto[face] = 1;
  proto.score = 1;
  return proto;
}

/**
 * Compose-with-face-unit: equivalent to a single `clinamenLift` on
 * `face`. The CostAlgebra perspective and the OperationalMesh
 * perspective agree. Mirrors
 * `Gnosis.CostAlgebra.buley_compose_with_face_unit_is_clinamen_lift`.
 */
export function composeWithFaceUnit(
  b: BuleyProjection,
  face: BuleyFace,
): BuleyProjection {
  return buleyCostAlgebra.compose(b, buleyFaceUnit(face));
}

/**
 * Validate that a candidate map preserves zero, composition, and score
 * for the given cost algebras — i.e. that it is a valid `CostHom`.
 * Returns a list of failing law names; empty list means the candidate
 * is a valid morphism on the witnesses provided.
 */
export function validateCostHom<S, T>(
  candidate: (s: S) => T,
  source: CostAlgebra<S>,
  target: CostAlgebra<T>,
  witnesses: readonly S[],
): readonly string[] {
  const failures: string[] = [];

  if (!target.equals(candidate(source.zero), target.zero)) {
    failures.push('preserve_zero');
  }

  for (const a of witnesses) {
    for (const b of witnesses) {
      const lhs = candidate(source.compose(a, b));
      const rhs = target.compose(candidate(a), candidate(b));
      if (!target.equals(lhs, rhs)) {
        failures.push(`preserve_compose@(${source.score(a)},${source.score(b)})`);
        break;
      }
    }
  }

  for (const s of witnesses) {
    if (target.score(candidate(s)) !== source.score(s)) {
      failures.push(`preserve_score@${source.score(s)}`);
    }
  }

  return failures;
}
