// @a0n/.ignored_gnosis/distributed-inference-host/src/deblur-public.ts
// Public surface for the deblur pipeline. Import from here in consumer code:
//
//   import {
//     deblurAnyImage, encodeRecoveredAsPNG,
//     forkRaceFoldDeblur,
//     attachDeblurEnsembleToMeshCarrierAssessmentJson,
//     meshCarrierRequestHeadersWithDeblurEnsemble,
//     deblurVideo,
//     deblurAudio, denoiseAudio,
//     bilateralFilter, bilateralFilterFRF,
//     selectSigmaByGoldenAttractor,
//     PHI, SQRT_5,
//   } from '@a0n/distributed-inference-host/deblur';
//
// Anything not re-exported here is considered internal — depend on it
// at your own risk; we may rename or refactor without warning.
//
// New 2026-04-27: moonshot-sieve exports — chromaticBuleSieve, bettiSignature,
// rejectionWeightedFold, crossModalSigmaScore, tritonStretch, plus the audio
// phoneme + pneuma bridges and the sieve-algebra (TS mirror of
// Gnosis/SubstrateSieve.lean).

export {
  deblurAnyImage,
  encodeRecoveredAsPNG,
  type DeblurAnyImageInput,
  type DeblurAnyImageOptions,
  type DeblurAnyImageResult,
  type DeblurChannelReport,
} from './deblur-any-image';
export {
  applyInvisibleWatermarkToPlane,
  resolveInvisibleWatermarkOptions,
  watermarkResidualCorrelation,
  type InvisibleWatermarkMode,
  type InvisibleWatermarkOptions,
  type InvisibleWatermarkProvenance,
  type InvisibleWatermarkResolved,
} from './invisible-watermark';

export {
  forkRaceFoldDeblur,
  type DeblurEnsembleArgs,
  type DeblurEnsembleConfig,
  type DeblurEnsembleResult,
  type DeblurEnsembleCandidateReport,
} from './forkraceFoldDeblur';

export {
  MESH_HTTP_HEADER_CARRIER_DEBLUR_FOLD,
  attachDeblurEnsembleToMeshCarrierAssessmentJson,
  deblurEnsembleResultToMeshCarrierFold,
  encodeDeblurFoldWireForMeshCarrierRequestHeader,
  mergeDeblurFoldIntoMeshCarrierJson,
  meshCarrierRequestHeadersWithDeblurEnsemble,
} from './mesh-carrier-assessment';

export {
  deblurVideo,
  type DeblurVideoArgs,
  type DeblurVideoResult,
} from './video-deblur';

export {
  deblurAudio,
  type DeblurAudioArgs,
  type DeblurAudioResult,
} from './audio-deblur';

export {
  denoiseAudio,
  type DenoiseAudioArgs,
  type DenoiseAudioResult,
} from './audio-denoise';

export {
  bilateralFilter,
  type BilateralFilterArgs,
} from './image-bilateral';

export {
  bilateralFilterFRF,
  type BilateralFRFArgs,
} from './image-bilateral-frf';

export {
  rgbFRF,
  splitRGB,
  joinRGB,
  type RGBFRFArgs,
  type ChannelTransform,
} from './image-rgb-frf';

export {
  selectSigmaByGoldenAttractor,
  selectSigmaByGoldenAttractorAsync,
  type SelectSigmaArgs,
  type SelectSigmaResult,
  type SigmaCandidateReport,
} from './sigma-attractor';

export {
  PHI,
  PHI_INVERSE,
  PHI_SQUARED,
  SQRT_5,
  PSI,
  binetFibonacci,
  binetResidual,
  goldenSpacedGrid,
  phiAttractorNewtonStep,
} from './attractor';

export {
  forkRaceFold,
  forkFold,
  type ForkTask,
  type ForkRaceFoldArgs,
  type ForkRaceFoldResult,
} from './frf';

export {
  FRFPool,
  withFRFPool,
  type WorkerFactory,
  type FRFPoolConfig,
  type PoolTask,
} from './frf-pool';

export {
  TOWER_CEILINGS,
  type BuleyFace,
  type BuleyProjection,
} from './spectral-noise-monitor';

export {
  SubstrateSieve,
  type Substrate,
} from './substrate-sieve';

// ---------------------------------------------------------------------------
// Moonshot sieves (topological reconstruction layer over the FRF substrate)
// ---------------------------------------------------------------------------

export {
  chromaticBuleSieve,
  ChromaticBuleSieve,
  type ChromaticBuleSieveArgs,
  type ChromaticBuleSieveResult,
  type ChromaticBuleSieveCellReport,
} from './chromatic-bule-sieve';

export {
  bettiSignature,
  isClosureAdmissible,
  signatureDistance,
  medianSignature,
  type BettiSignature,
  type BettiSignatureArgs,
} from './betti-signature';

export {
  rejectionWeightedFold,
  type RejectionPoint,
  type SurvivingCandidate,
  type RejectionWeightArgs,
  type RejectionWeightResult,
} from './buleyean-rejection-weights';

export {
  crossModalSigmaScore,
  type CrossModalScoreArgs,
  type CrossModalReport,
  type CrossModalSequenceInput,
} from './cross-modal-moonshine';

export {
  tritonStretch,
  type TritonStretchArgs,
  type TritonStretchResult,
  type TritonStretchLevel,
} from './triton-stretch';

// ---------------------------------------------------------------------------
// Audio: phoneme + pneuma codec bridge
// ---------------------------------------------------------------------------

export {
  extractPhonemeFingerprints,
  BAND_EDGES_HZ,
  type PhonemeFingerprint,
  type PhonemeFingerprintArgs,
  type PhonemeFingerprintResult,
  type PhonemeCategory,
} from './phoneme-fingerprint';

export {
  recoverAudioViaPneuma,
  type PneumaAudioRecoverArgs,
  type PneumaAudioRecoverResult,
} from './audio-pneuma-bridge';

// ---------------------------------------------------------------------------
// Sieve algebra (TS mirror of Gnosis/SubstrateSieve.lean)
// ---------------------------------------------------------------------------
// NOTE: `Substrate` is intentionally NOT re-exported from `./sieve-algebra`
// here — the name is already taken by `./substrate-sieve` (concrete record),
// and sieve-algebra's `Substrate<A>` is a generic envelope. Consumers who
// need the generic form should import it directly from the source module.

export {
  identitySieve,
  verifySieveLaws,
  TIER_BANDWIDTHS,
  TIERS,
  type Tier,
  type SieveOps,
} from './sieve-algebra';

// ---------------------------------------------------------------------------
// FFT FRF (the structural slot for the future WASM SIMD migration)
// ---------------------------------------------------------------------------

export {
  fft2FRF,
  ifft2FRF,
  type FFT2FRFArgs,
  type IFFT2FRFArgs,
} from './image-fft-frf';
