// @a0n/.ignored_gnosis/distributed-inference-host/src/direction-of-arrival.ts
/**
 * # Direction of Arrival — a bearing is admitted only when the array agrees
 *
 * Mirrors `Gnosis.Mesh.DirectionOfArrival` from `open-source/gnosis-math`, the
 * Rustic Church carrier for the Monster Mesh direction-of-arrival (DOA) surface.
 * The physical picture:
 *
 *  - A mesh of microphones at known positions hears one incident sound. Because
 *    the sound reaches the microphones at slightly different times, the *moveout*
 *    — the way arrival time changes with microphone position — encodes the
 *    **bearing** (direction) the sound came from. This is the beamforming / DOA
 *    principle.
 *  - In the far field the incident wavefront is planar, so along a 1-D array the
 *    arrival time is a **linear** function of microphone position: the arrival at
 *    position `p` is `refArrival + slowness * p`, where `slowness` is the
 *    per-unit arrival increment (the projection of the wave's slowness vector
 *    onto the array axis) and `refArrival` is the arrival at position 0. The
 *    runtime geometry layer recovers the physical angle `θ = arcsin(slowness·c)`;
 *    that trigonometry is OUTSIDE this carrier.
 *
 * ## What is (and is not) carried
 *
 * This module is the runtime promotion check for a DOA bearing. It carries ONLY
 * the finite contract the Lean file proves:
 *
 *  - the far-field linear moveout prediction (`predictedArrival`);
 *  - the arrival residual against a bearing's predicted moveout (`residual`);
 *  - a detected microphone's consistency at a tolerance (`consistent`);
 *  - the count of mutually-consistent microphones (`countConsistent`);
 *  - the quorum gate: a bearing needs `K >= 2` consistent microphones, because
 *    one arrival time is consistent with *every* bearing (it constrains only
 *    `refArrival`, not the slope) — a lone ear cannot beamform
 *    (`single_node_no_bearing`);
 *  - the spoof rejection: a microphone whose fabricated arrival contradicts the
 *    established moveout is inconsistent and cannot inflate the count
 *    (`spoof_node_inconsistent` / `spoof_does_not_inflate`).
 *
 * True clock synchronization, real 2-D/3-D angle recovery via `arcsin`, sound-
 * speed calibration, near-field curvature, and multipath stay runtime/empirical
 * obligations OUTSIDE this certificate — they are not theorems and are not
 * asserted here.
 *
 * The promotion check folds the contract into the same epistemic-outcome
 * vocabulary the sibling mesh carriers already speak
 * (`claim_admitted` / `observation_only` / `claim_withheld`), so a bearing
 * cannot be displayed as a resolved direction until a quorum of independent
 * microphones has cleared the Lean-side moveout/quorum gate.
 *
 * ## Lean source-of-truth
 *
 * @see open-source/gnosis-math/Gnosis/Mesh/DirectionOfArrival.lean
 * @see open-source/gnosis-math/Gnosis/Mesh/CrossNodeCoincidence.lean (mesh sibling)
 */

/* ------------------------------------------------------------------ */
/*  Bearings, microphone nodes, and predicted arrivals                */
/* ------------------------------------------------------------------ */

/**
 * A microphone in the array: its position on the 1-D axis, the tick it reported
 * hearing the event, and whether it detected anything at all. Positions and
 * arrivals are non-negative integers (saturating `Nat` arithmetic), matching the
 * Lean `Node`.
 */
export interface MeshNode {
  readonly position: number;
  readonly reportedArrival: number;
  readonly detected: boolean;
}

/**
 * A candidate incident direction, modeled by its far-field linear moveout: the
 * arrival at position 0 (`refArrival`) and the per-position arrival increment
 * (`slowness`, the projection of the slowness vector onto the array axis). The
 * physical angle `arcsin(slowness·c)` is recovered at runtime.
 * Mirrors `DirectionOfArrival.Bearing`.
 */
export interface Bearing {
  readonly slowness: number;
  readonly refArrival: number;
}

/** Absolute difference on `Nat` via saturating subtraction (`|a - b|`). The
 *  arrival residual metric, shared with the mesh corridor carriers.
 *  Mirrors `DirectionOfArrival.adist`. */
export function adist(a: number, b: number): number {
  return Math.max(0, a - b) + Math.max(0, b - a);
}

/** Predicted arrival of a bearing at a position under the far-field plane-wave
 *  model: `refArrival + slowness * position`.
 *  Mirrors `DirectionOfArrival.predictedArrival`. */
export function predictedArrival(b: Bearing, pos: number): number {
  return b.refArrival + b.slowness * pos;
}

/** The arrival residual: how far a microphone's reported arrival is from the
 *  arrival this bearing predicts at the microphone's position.
 *  Mirrors `DirectionOfArrival.residual`. */
export function residual(b: Bearing, node: MeshNode): number {
  return adist(node.reportedArrival, predictedArrival(b, node.position));
}

/* ------------------------------------------------------------------ */
/*  Consistency and quorum admission                                  */
/* ------------------------------------------------------------------ */

/** A microphone is consistent with a bearing when it detected the event and its
 *  arrival residual clears the tolerance.
 *  Mirrors `DirectionOfArrival.consistentB`. */
export function consistent(tol: number, b: Bearing, node: MeshNode): boolean {
  return node.detected && residual(b, node) <= tol;
}

/** How many microphones in a list are consistent with the bearing.
 *  Mirrors `DirectionOfArrival.countConsistent`. */
export function countConsistent(
  tol: number,
  b: Bearing,
  nodes: readonly MeshNode[],
): number {
  let count = 0;
  for (const node of nodes) {
    if (consistent(tol, b, node)) {
      count += 1;
    }
  }
  return count;
}

/* ------------------------------------------------------------------ */
/*  Runtime promotion check (epistemic outcome)                       */
/* ------------------------------------------------------------------ */

/**
 * Epistemic outcome of a DOA bearing observation. Identical vocabulary to the
 * sibling mesh carriers (`./cross-node-coincidence`,
 * `./electrical-network-frequency`), kept as a local literal union so this
 * certificate has no import coupling:
 *
 *  - `claim_admitted`    — a quorum `K >= 2` of independent microphones are
 *                          mutually consistent with one bearing's moveout, so the
 *                          bearing may be promoted to a resolved-direction claim.
 *  - `observation_only`  — one or more microphones are consistent but the quorum
 *                          is not met (an unquorate bearing hint), OR the
 *                          requested quorum is `K < 2` and so cannot resolve a
 *                          direction (one arrival fits every bearing); the
 *                          observation is reported, never promoted.
 *  - `claim_withheld`    — no microphone is consistent with the bearing, so there
 *                          is no direction to claim.
 */
export type DirectionOfArrivalEpistemicOutcome =
  | 'claim_admitted'
  | 'observation_only'
  | 'claim_withheld';

export interface DirectionOfArrivalAdmission {
  readonly outcome: DirectionOfArrivalEpistemicOutcome;
  /** Indices (into the node list) of the microphones consistent with the bearing. */
  readonly consistentNodes: number[];
  /** How many microphones are mutually consistent with the bearing. */
  readonly consistentCount: number;
  /** Whether the quorum was met (count >= K AND K >= 2). */
  readonly quorumMet: boolean;
  readonly summary: string;
}

/**
 * The promotion check Monster Mesh runs before any UI surface displays a DOA
 * bearing as a resolved direction. It folds the Lean contract into an epistemic
 * outcome: the consistency count decides whether there is anything to claim, and
 * the quorum `K >= 2` decides whether the claim is allowed to be a *resolved-
 * direction* claim rather than the report of a single microphone whose lone
 * arrival is consistent with every bearing.
 *
 * A quorum of `K < 2` is refused outright (mirrors how the Lean module forbids
 * quorum-1 admission via `single_node_no_bearing`): one arrival time fixes only
 * `refArrival` and constrains no slope, so a lone microphone never resolves a
 * direction and never reaches a promotable claim.
 *
 * @param nodes   the microphones observed for this candidate bearing
 * @param bearing the moveout (slowness + reference arrival) being tested
 * @param quorumK the number of mutually-consistent microphones required to
 *                promote (must be >= 2 to inherit the single-node refutation)
 * @param tol     the arrival-residual tolerance
 */
export function admitBearing(
  nodes: readonly MeshNode[],
  bearing: Bearing,
  quorumK: number,
  tol: number,
): DirectionOfArrivalAdmission {
  const consistentNodes: number[] = [];
  for (let i = 0; i < nodes.length; i++) {
    if (consistent(tol, bearing, nodes[i])) {
      consistentNodes.push(i);
    }
  }
  const consistentCount = consistentNodes.length;

  // A quorum below two cannot inherit the Lean single-node refutation
  // (`single_node_no_bearing` forbids admission at K = 1): one arrival is
  // consistent with every bearing, so the gate refuses to resolve a direction.
  const quorumIsSane = quorumK >= 2;
  const quorumMet = quorumIsSane && consistentCount >= quorumK;

  if (consistentCount === 0) {
    return {
      outcome: 'claim_withheld',
      consistentNodes,
      consistentCount,
      quorumMet: false,
      summary:
        'no microphone is consistent with the bearing — no direction to claim',
    };
  }

  if (quorumMet) {
    return {
      outcome: 'claim_admitted',
      consistentNodes,
      consistentCount,
      quorumMet: true,
      summary: `quorum K=${quorumK} met: ${consistentCount} independent microphones agree on the moveout slope`,
    };
  }

  return {
    outcome: 'observation_only',
    consistentNodes,
    consistentCount,
    quorumMet: false,
    summary: quorumIsSane
      ? `unquorate bearing hint: ${consistentCount} consistent microphone(s) below quorum K=${quorumK}`
      : `quorum K=${quorumK} < 2 cannot resolve a direction from a single arrival`,
  };
}

/* ------------------------------------------------------------------ */
/*  Concrete array fixture (mirrors the Lean fixtures)                */
/* ------------------------------------------------------------------ */

/** The true incident bearing: moveout slope 3, reference arrival 100. So the
 *  predicted arrivals are 100 (pos 0), 106 (pos 2), 112 (pos 4).
 *  Mirrors `DirectionOfArrival.trueBearing`. */
export const trueBearing: Bearing = { slowness: 3, refArrival: 100 };

/** The wrong (broadside) bearing: slope 0, so it predicts arrival 100
 *  everywhere. It explains the position-0 microphone but no other.
 *  Mirrors `DirectionOfArrival.broadsideBearing`. */
export const broadsideBearing: Bearing = { slowness: 0, refArrival: 100 };

/** Microphone at position 0: true arrival `100 + 3*0 = 100`, reported exactly.
 *  Mirrors `DirectionOfArrival.nodeA`. */
export const nodeA: MeshNode = {
  position: 0,
  reportedArrival: 100,
  detected: true,
};

/** Microphone at position 2: true arrival `100 + 3*2 = 106`, reported exactly.
 *  Mirrors `DirectionOfArrival.nodeB`. */
export const nodeB: MeshNode = {
  position: 2,
  reportedArrival: 106,
  detected: true,
};

/** Microphone at position 4: true arrival `100 + 3*4 = 112`, reported exactly.
 *  Mirrors `DirectionOfArrival.nodeC`. */
export const nodeC: MeshNode = {
  position: 4,
  reportedArrival: 112,
  detected: true,
};

/** A spoof microphone at position 6 claiming it heard the event at tick 100. The
 *  true bearing predicts `100 + 3*6 = 118` there, so its residual is
 *  `|100 - 118| = 18`, far beyond tolerance 1. Rejected.
 *  Mirrors `DirectionOfArrival.spoofNode`. */
export const spoofNode: MeshNode = {
  position: 6,
  reportedArrival: 100,
  detected: true,
};

/** The three honest microphones. Mirrors `DirectionOfArrival.realNodes`. */
export const realNodes: readonly MeshNode[] = [nodeA, nodeB, nodeC];

/**
 * Direction-of-Arrival master witness. Bundles the boundary facts the Lean
 * theorem `doa_master` proves, each recomputed here from first principles
 * (never imported from Lean IR).
 *
 * Mirrors `DirectionOfArrival.doa_master`.
 */
export const DirectionOfArrivalMaster = {
  /** The honest array confirms the true bearing at a quorum of two. */
  trueBearingAdmittedAtTwo:
    admitBearing(realNodes, trueBearing, 2, 1).outcome === 'claim_admitted',

  /** The honest agreement is full: all three microphones confirm the bearing. */
  trueBearingAdmittedAtThree:
    admitBearing(realNodes, trueBearing, 3, 1).outcome === 'claim_admitted',

  /** The wrong broadside bearing (flat slope) reaches only the position-0
   *  microphone, so it cannot reach the quorum of two and is not admitted. */
  broadsideNotAdmitted:
    admitBearing(realNodes, broadsideBearing, 2, 1).outcome !==
    'claim_admitted',

  /** A single honest microphone, alone, is not admitted at quorum two. */
  loneNodeNotAdmitted:
    admitBearing([nodeA], trueBearing, 2, 1).outcome !== 'claim_admitted',

  /** Distinct slownesses give distinct arrivals at any nonzero position: the
   *  moveout slope is the discriminant (`106 != 100` at position 2). */
  bearingsDistinctOfSlowness:
    predictedArrival(trueBearing, 2) !==
    predictedArrival(broadsideBearing, 2),

  /** The spoof microphone is inconsistent with the bearing the honest array
   *  confirms: its arrival contradicts the established moveout. */
  spoofNodeInconsistent: consistent(1, trueBearing, spoofNode) === false,

  /** The spoof cannot inflate the consistency count: count over the honest array
   *  plus the spoof equals the count over the honest array alone. */
  spoofDoesNotInflate:
    countConsistent(1, trueBearing, [...realNodes, spoofNode]) ===
    countConsistent(1, trueBearing, realNodes),

  /** Same reference arrival, opposite verdicts: the slope discriminates, so a
   *  reference arrival alone cannot be used as a bearing key. */
  sameRefOppositeVerdict:
    trueBearing.refArrival === broadsideBearing.refArrival &&
    admitBearing(realNodes, trueBearing, 2, 1).outcome === 'claim_admitted' &&
    admitBearing(realNodes, broadsideBearing, 2, 1).outcome !==
      'claim_admitted',
} as const;
