// @a0n/.ignored_gnosis/distributed-inference-host/src/doppler-speed.ts
/**
 * # Doppler Speed — the pass-through sweep as a radial-speed admission gate
 *
 * Mirrors `Gnosis.Mesh.DopplerSpeed` from `open-source/gnosis-math`, the Rustic
 * Church carrier for the Doppler-speed idea: a source moving past a microphone
 * produces a frequency *sweep*. While the source approaches, the observed
 * frequency is shifted up; while it recedes, it is shifted down. Through a single
 * clean pass the observed frequency falls monotonically from a high (approaching)
 * plateau to a low (receding) plateau, and the radial speed is read from the
 * *total drop* across the pass.
 *
 * ## What is (and is not) carried
 *
 * This module is the runtime promotion check for a Doppler-speed observation. It
 * carries ONLY the finite contract the Lean file proves, two gates:
 *
 *  - the monotone gate (the single-pass consistency analog): a single pass is
 *    monotone non-increasing. A sequence that rises and falls is noise, jitter,
 *    or several independent sources — not one pass-through
 *    (`monotoneNonIncreasing`);
 *  - the drop / support gate: the total drop (head minus last) must exceed
 *    `minDrop`. A flat sweep has drop zero — a stationary source modulates no
 *    sweep, no resolvable radial motion (`totalDrop`).
 *
 * Real STFT frequency tracking, the Doppler equation
 * `f_obs = f_src * c / (c + v_radial)`, the carrier-speed and emitted-frequency
 * calibration, multi-source separation, and SNR robustness stay runtime/
 * empirical obligations OUTSIDE this certificate — they are not theorems and are
 * not asserted here.
 *
 * The promotion check folds the contract into the same epistemic-outcome
 * vocabulary the sibling mesh carriers already speak
 * (`claim_admitted` / `observation_only` / `claim_withheld`), so a Doppler-speed
 * observation cannot be displayed as a speed claim until it has cleared the
 * Lean-side monotone/drop gate. The hard negative is the *same source in two
 * motion states*: a moving source (monotone, large drop) and a stationary source
 * (flat, zero drop) are one emitter, and only the sweep distinguishes them —
 * source identity alone cannot supply a speed.
 *
 * ## Lean source-of-truth
 *
 * @see open-source/gnosis-math/Gnosis/Mesh/DopplerSpeed.lean
 * @see open-source/gnosis-math/Gnosis/Mesh/VitalSignMicroDoppler.lean (motion sibling)
 * @see open-source/gnosis-math/Gnosis/Mesh/ElectricalNetworkFrequency.lean (variation sibling)
 */

/* ------------------------------------------------------------------ */
/*  Observed-frequency sweeps                                         */
/* ------------------------------------------------------------------ */

/**
 * An observed sweep is a list of per-frame observed frequencies, ordered through
 * a pass (nominally high while approaching, low while receding). Frequencies are
 * non-negative integers (saturating `Nat` arithmetic), matching the Lean
 * `Sweep`.
 */
export type Sweep = readonly number[];

/* ------------------------------------------------------------------ */
/*  Monotone gate                                                     */
/* ------------------------------------------------------------------ */

/**
 * A sweep is monotone non-increasing when each element is at least the next: the
 * signature of one clean pass. The empty and singleton sweeps are vacuously
 * monotone. Mirrors `DopplerSpeed.monotoneNonIncreasingB`.
 */
export function monotoneNonIncreasing(freqs: Sweep): boolean {
  for (let i = 1; i < freqs.length; i++) {
    if (freqs[i] > freqs[i - 1]) {
      return false;
    }
  }
  return true;
}

/* ------------------------------------------------------------------ */
/*  Total drop (the speed proxy)                                      */
/* ------------------------------------------------------------------ */

/**
 * The total drop across the sweep: head minus last (saturating). For a clean
 * descending pass this is the high-plateau frequency minus the low-plateau
 * frequency, the proxy for radial speed. A flat sweep drops to `0`; an empty
 * sweep has drop `0`. Mirrors `DopplerSpeed.totalDrop`.
 */
export function totalDrop(freqs: Sweep): number {
  if (freqs.length === 0) {
    return 0;
  }
  const head = freqs[0];
  const last = freqs[freqs.length - 1];
  return Math.max(0, head - last);
}

/* ------------------------------------------------------------------ */
/*  Runtime promotion check (epistemic outcome)                       */
/* ------------------------------------------------------------------ */

/**
 * Epistemic outcome of a Doppler-speed observation. Identical vocabulary to the
 * sibling mesh carriers (`./vital-sign-micro-doppler`,
 * `./electrical-network-frequency`), kept as a local literal union so this
 * certificate has no import coupling:
 *
 *  - `claim_admitted`    — a monotone sweep whose total drop clears `minDrop`,
 *                          so the radial speed (proportional to the drop) may be
 *                          promoted to a claim.
 *  - `claim_withheld`    — the sweep is non-monotone (erratic — no clean pass)
 *                          OR it is monotone but its drop is below `minDrop` (a
 *                          stationary / flat source with no resolvable radial
 *                          motion), so there is no speed to claim.
 *
 * (`observation_only` is part of the shared vocabulary but is not reached by the
 *  two-gate Doppler-speed contract: every non-admitting sweep is withheld, as
 *  the Lean module proves both the erratic and stationary fixtures are rejected
 *  outright.)
 */
export type DopplerSpeedEpistemicOutcome =
  | 'claim_admitted'
  | 'observation_only'
  | 'claim_withheld';

/** The reason a Doppler-speed claim was held back from promotion. */
export type DopplerSpeedLimitationClass =
  | 'not_one_clean_pass'
  | 'insufficient_drop'
  | 'none';

export interface DopplerSpeedAdmission {
  readonly outcome: DopplerSpeedEpistemicOutcome;
  readonly limitation: DopplerSpeedLimitationClass;
  /** Whether the sweep is monotone non-increasing (one clean pass). */
  readonly monotone: boolean;
  /** The measured total drop (the displayable speed proxy). */
  readonly totalDrop: number;
  readonly summary: string;
}

export interface DopplerSpeedAdmitOptions {
  /** The minimum total drop required to read a resolvable radial speed. */
  readonly minDrop: number;
}

/**
 * The promotion check Monster Mesh runs before any UI surface displays a
 * Doppler-speed observation as a speed claim. It folds the Lean contract into an
 * epistemic outcome:
 *
 *  - the monotone gate decides whether the sweep is one clean pass — an erratic
 *    rise-and-fall is noise / multiple sources and is withheld (mirroring
 *    `erratic_rejected`);
 *  - the drop gate decides whether there is resolvable radial motion — a
 *    monotone sweep whose drop is below `minDrop` is a stationary / flat source
 *    and is withheld (mirroring `stationary_rejected` and the general
 *    `flat_not_admitted` lemma).
 *
 * Only a monotone sweep with sufficient drop is admitted; the speed is read as
 * proportional to that drop.
 *
 * @param freqs the observed-frequency sweep under examination
 * @param opts  the minimum total drop floor
 */
export function admitSpeed(
  freqs: Sweep,
  opts: DopplerSpeedAdmitOptions,
): DopplerSpeedAdmission {
  const monotone = monotoneNonIncreasing(freqs);
  const drop = totalDrop(freqs);

  // Monotone gate: a non-monotone sweep is not one clean pass — there is no
  // single-source sweep to read a speed from.
  if (!monotone) {
    return {
      outcome: 'claim_withheld',
      limitation: 'not_one_clean_pass',
      monotone,
      totalDrop: drop,
      summary:
        'non-monotone sweep: rises and falls — not one clean pass, no speed to read',
    };
  }

  // Drop gate: a monotone but low-drop sweep is a stationary / flat source — no
  // resolvable radial motion.
  if (drop < opts.minDrop) {
    return {
      outcome: 'claim_withheld',
      limitation: 'insufficient_drop',
      monotone,
      totalDrop: drop,
      summary: `monotone but drop ${drop} below floor ${opts.minDrop} — stationary source, no resolvable speed`,
    };
  }

  return {
    outcome: 'claim_admitted',
    limitation: 'none',
    monotone,
    totalDrop: drop,
    summary: `one clean pass with drop ${drop} >= ${opts.minDrop} — radial speed read (proportional to drop)`,
  };
}

/* ------------------------------------------------------------------ */
/*  Finite certificate                                                */
/* ------------------------------------------------------------------ */

/**
 * A finite certificate that Monster Mesh has all required Doppler-speed gates.
 * Mirrors `DopplerSpeed.DopplerSpeedCertificate`.
 */
export interface DopplerSpeedCertificate {
  readonly sweepModeled: boolean;
  readonly monotoneGated: boolean;
  readonly dropSupported: boolean;
  readonly stationaryRejected: boolean;
  readonly erraticRejected: boolean;
}

/** The Doppler-speed carrier for Monster Mesh promotion checks.
 *  Mirrors `DopplerSpeed.dopplerSpeedRusticCertificate`. */
export const dopplerSpeedRusticCertificate: DopplerSpeedCertificate = {
  sweepModeled: true,
  monotoneGated: true,
  dropSupported: true,
  stationaryRejected: true,
  erraticRejected: true,
};

/* ------------------------------------------------------------------ */
/*  Concrete fixtures (mirror the Lean fixtures)                      */
/* ------------------------------------------------------------------ */

/** A source moving past the mic: a clean descending sweep. Approaching plateau
 *  120 falls through the pass to receding plateau 70 — drop 50. Mirrors `moving`. */
export const moving: Sweep = [120, 110, 95, 80, 70];

/** The same source, stationary: a flat observed frequency. No sweep, no speed.
 *  drop 0. Mirrors `stationary`. */
export const stationary: Sweep = [100, 100, 100, 100];

/** An erratic sequence: the observed frequency rises (100 → 120) then falls —
 *  not one clean pass. Noise, jitter, or multiple sources. Mirrors `erratic`. */
export const erratic: Sweep = [100, 120, 90, 130];

/**
 * Doppler-Speed master witness. Bundles the boundary facts the Lean theorem
 * `doppler_master` proves, each recomputed here from first principles (never
 * imported from Lean IR). Gate parameter: minDrop 10.
 *
 * Mirrors `DopplerSpeed.doppler_master`.
 */
export const DopplerSpeedMaster = {
  /** A moving source is admitted: one clean pass with drop 50 >= 10. */
  movingAdmitted:
    admitSpeed(moving, { minDrop: 10 }).outcome === 'claim_admitted',

  /** A stationary source is rejected: the sweep is flat, drop 0 < 10. */
  stationaryRejected:
    admitSpeed(stationary, { minDrop: 10 }).outcome !== 'claim_admitted',

  /** An erratic sequence is rejected: it is not one clean pass. */
  erraticRejected:
    admitSpeed(erratic, { minDrop: 10 }).outcome !== 'claim_admitted',

  /** The moving sweep is monotone non-increasing. */
  movingMonotone: monotoneNonIncreasing(moving) === true,

  /** The erratic sweep is NOT monotone (rises at 100 → 120). */
  erraticNotMonotone: monotoneNonIncreasing(erratic) === false,

  /** The moving sweep's total drop is 50 (head 120 minus last 70). */
  movingDrop: totalDrop(moving) === 50,

  /** The stationary sweep's total drop is 0 (head 100 minus last 100). */
  stationaryDrop: totalDrop(stationary) === 0,

  /** A constant sweep of positive length has drop 0 (general flat lemma). */
  flatHasZeroDrop: totalDrop([100, 100, 100]) === 0,

  /** The hard negative: same source, moving admitted, stationary rejected. */
  movingVsStationary:
    admitSpeed(moving, { minDrop: 10 }).outcome === 'claim_admitted' &&
    admitSpeed(stationary, { minDrop: 10 }).outcome !== 'claim_admitted',

  /** The certificate records stationary rejection. */
  certStationaryRejected: dopplerSpeedRusticCertificate.stationaryRejected,

  /** The certificate records erratic rejection. */
  certErraticRejected: dopplerSpeedRusticCertificate.erraticRejected,
} as const;
