// @a0n/.ignored_gnosis/distributed-inference-host/src/fft2.ts
/**
 * 2D radix-2 Cooley-Tukey FFT for image-carrier deblur.
 *
 * Status: hot path for `image-deblur.ts` and `image-psf.ts`. Self-contained
 * complex input/output; deliberately not built on top of the real-input
 * `fft` in `mel-spectrogram.ts` because the column pass needs complex
 * input and forcing that through the audio API would mean two passes per
 * column or reshaping the audio function. The audio FFT stays untouched.
 *
 * Power-of-2 dimensions only; asserts loud (memory: silent catch is silent
 * loss). For our smoke fixtures (256x256, 512x512) this is never a constraint.
 */

export interface Complex2D {
  readonly re: Float32Array;
  readonly im: Float32Array;
  readonly w: number;
  readonly h: number;
}

/**
 * Returns whether is Power Of Two is true.
 */
export function isPowerOfTwo(n: number): boolean {
  return Number.isInteger(n) && n > 0 && (n & (n - 1)) === 0;
}

function assertPow2(n: number, label: string): void {
  if (!isPowerOfTwo(n)) {
    throw new RangeError(`${label} must be a positive power of two; got ${n}`);
  }
}

/**
 * In-place 1D complex FFT. `sign = -1` for forward, `sign = +1` for
 * inverse (caller normalizes by 1/n on inverse).
 */
function fft1dComplex(re: Float32Array, im: Float32Array, n: number, sign: number): void {
  // Bit-reversal permutation
  let j = 0;
  for (let i = 0; i < n - 1; i++) {
    if (i < j) {
      const tr = re[i]!, ti = im[i]!;
      re[i] = re[j]!; im[i] = im[j]!;
      re[j] = tr; im[j] = ti;
    }
    let m = n >> 1;
    while (m >= 2 && j >= m) { j -= m; m >>= 1; }
    j += m;
  }
  // Butterfly
  const stages = Math.log2(n);
  for (let stage = 1; stage <= stages; stage++) {
    const m = 1 << stage;
    const halfM = m >> 1;
    const angle = (sign * 2 * Math.PI) / m;
    const wpr = Math.cos(angle);
    const wpi = Math.sin(angle);
    for (let k = 0; k < n; k += m) {
      let wr = 1.0, wi = 0.0;
      for (let s = 0; s < halfM; s++) {
        const a = k + s;
        const b = a + halfM;
        const ar = re[a]!, ai = im[a]!;
        const br = re[b]!, bi = im[b]!;
        const tr = wr * br - wi * bi;
        const ti = wr * bi + wi * br;
        re[b] = ar - tr;
        im[b] = ai - ti;
        re[a] = ar + tr;
        im[a] = ai + ti;
        const wrNext = wr * wpr - wi * wpi;
        wi = wr * wpi + wi * wpr;
        wr = wrNext;
      }
    }
  }
}

function rowColPass(re: Float32Array, im: Float32Array, w: number, h: number, sign: number): void {
  // Row pass
  for (let y = 0; y < h; y++) {
    const off = y * w;
    fft1dComplex(re.subarray(off, off + w), im.subarray(off, off + w), w, sign);
  }
  // Column pass: gather → FFT → scatter
  const colRe = new Float32Array(h);
  const colIm = new Float32Array(h);
  for (let x = 0; x < w; x++) {
    for (let y = 0; y < h; y++) { colRe[y] = re[y * w + x]!; colIm[y] = im[y * w + x]!; }
    fft1dComplex(colRe, colIm, h, sign);
    for (let y = 0; y < h; y++) { re[y * w + x] = colRe[y]!; im[y * w + x] = colIm[y]!; }
  }
}

/**
 * Forward 2D FFT on real input. Returns complex spectrum of size w*h.
 * Output layout matches the input (no fftshift); DC is at index 0.
 */
export function fft2(real: Float32Array, w: number, h: number): Complex2D {
  assertPow2(w, 'fft2: w');
  assertPow2(h, 'fft2: h');
  if (real.length !== w * h) {
    throw new RangeError(`fft2: real.length ${real.length} != w*h ${w * h}`);
  }
  const re = new Float32Array(real);
  const im = new Float32Array(w * h);
  rowColPass(re, im, w, h, -1);
  return { re, im, w, h };
}

/**
 * Forward 2D FFT on complex input. The arrays are copied; caller's buffers
 * are untouched.
 */
export function fft2Complex(re: Float32Array, im: Float32Array, w: number, h: number): Complex2D {
  assertPow2(w, 'fft2Complex: w');
  assertPow2(h, 'fft2Complex: h');
  if (re.length !== w * h || im.length !== w * h) {
    throw new RangeError(`fft2Complex: arrays must be length w*h`);
  }
  const reCopy = new Float32Array(re);
  const imCopy = new Float32Array(im);
  rowColPass(reCopy, imCopy, w, h, -1);
  return { re: reCopy, im: imCopy, w, h };
}

/**
 * Inverse 2D FFT. Returns the real part of the inverse transform; the
 * imaginary part should be ~0 if the spectrum was conjugate-symmetric
 * (i.e. the forward input was real). Normalized by 1/(w*h).
 */
export function ifft2(c: Complex2D): Float32Array {
  const { w, h } = c;
  assertPow2(w, 'ifft2: w');
  assertPow2(h, 'ifft2: h');
  const re = new Float32Array(c.re);
  const im = new Float32Array(c.im);
  rowColPass(re, im, w, h, +1);
  const norm = 1.0 / (w * h);
  const out = new Float32Array(w * h);
  for (let i = 0; i < re.length; i++) out[i] = re[i]! * norm;
  return out;
}

/**
 * Inverse 2D FFT returning the full complex result (both parts), normalized
 * by 1/(w*h). Useful when the spectrum is *not* conjugate-symmetric (e.g.
 * after a Wiener step on a half-truncated kernel).
 */
export function ifft2Complex(c: Complex2D): Complex2D {
  const { w, h } = c;
  assertPow2(w, 'ifft2Complex: w');
  assertPow2(h, 'ifft2Complex: h');
  const re = new Float32Array(c.re);
  const im = new Float32Array(c.im);
  rowColPass(re, im, w, h, +1);
  const norm = 1.0 / (w * h);
  for (let i = 0; i < re.length; i++) { re[i] = re[i]! * norm; im[i] = im[i]! * norm; }
  return { re, im, w, h };
}
