// @a0n/.ignored_gnosis/distributed-inference-host/src/forkraceFoldDeblur.ts
// @ts-nocheck
// File header (7-9 lines, MANDATORY): the FRF ensemble for image deblur.
// Mirrors forkRaceFoldMOA from @a0n/aether/src/glossolalia-moa.ts — same
// Buleyean deficit-weighted merge applied to image planes instead of token
// logits. FORK runs N deblur configs in parallel. RACE filters out configs
// whose carrierBuley.score violates the tower ceiling. FOLD does a pixel-
// wise weighted average where each candidate's weight is buleyeanWeight(R, v)
// with v = its rank by |Binet residual| from the phi-attractor (lower
// residual = fewer rejections = more weight). The winner's pixel anchors;
// peers blend in proportion to their attractor proximity. Two optional
// sieves layer on top: `bettiSieve` rejects survivors whose recovered plane
// has sum(beta) > closure ceiling (Pleromatic Closure), and
// `rejectionWeights` re-weights survivors via sigma-proximity to rejected
// candidates instead of the plain rank weight. Both off by default.

import { buleyeanWeight } from '@a0n/buleyean-kernel';
import { forkRaceFold, type ForkTask } from './frf';
import {
  deblurAnyImage,
  type DeblurAnyImageInput,
  type DeblurAnyImageOptions,
  type DeblurAnyImageResult,
} from './deblur-any-image';
import { TOWER_CEILINGS } from './spectral-noise-monitor';
import { bettiSignature, isClosureAdmissible } from './betti-signature';
import {
  rejectionWeightedFold,
  type RejectionPoint,
  type SurvivingCandidate,
} from './buleyean-rejection-weights';
import {
  applyInvisibleWatermarkToPlane,
  type InvisibleWatermarkOptions,
  type InvisibleWatermarkProvenance,
} from './invisible-watermark';

export interface DeblurEnsembleConfig {
  /** Free-form label for the report. */
  label?: string;
  /** Forwarded to deblurAnyImage. */
  options: DeblurAnyImageOptions;
}

export interface DeblurEnsembleArgs {
  input: DeblurAnyImageInput;
  configs: ReadonlyArray<DeblurEnsembleConfig>;
  /** Concurrency cap for the FORK step. Default: configs.length. */
  concurrency?: number;
  /** Tower ceiling on carrierBuley.score. Default TOWER_CEILINGS.hexon. */
  ceiling?: number;
  /** When set, also filter survivors by Betti closure: a survivor whose
   *  recovered plane has sum(beta) > ceiling is rejected. Default off
   *  (no Betti sieve). */
  bettiSieve?: {
    /** Total-β closure ceiling. Default 10 (TOWER_CEILINGS.decagon). */
    ceiling?: number;
    /** Threshold fraction for super-level-set computation. Default 0.5. */
    thresholdFraction?: number;
  };
  /** When set, weight survivors via the proximity-aware Buleyean formula
   *  using the rejected candidates' σ values as a prior. The default fold
   *  uses plain `buleyeanWeight(R, rank)`. */
  rejectionWeights?: {
    /** Proximity influence strength in [0, 1]. Default 0.5. */
    alpha?: number;
  };
  /** Optional invisible watermark applied once to the final folded plane. */
  invisibleWatermark?: InvisibleWatermarkOptions;
}

export interface DeblurEnsembleCandidateReport {
  label: string;
  selectedSigma: number;
  methodUsed: 'wiener' | 'richardson-lucy' | 'mesh-residual';
  binetResidual: number; // 0 if no attractor was run
  maxChannelScore: number; // max carrierBuley.score across channels
  rank: number; // 0 = best, increasing
  buleyWeight: number; // buleyeanWeight(R, rank)
  rejectedByCeiling: boolean;
  /** When the Betti sieve removed this candidate, this is true AND
   *  rejectedByCeiling is also true (it is the umbrella "removed" flag).
   *  Undefined when the Betti sieve was not enabled. */
  rejectedByBettiClosure?: boolean;
  /** Sum of (β0 + β1 + β2) on the luminance plane when the Betti sieve
   *  was enabled. Undefined otherwise. */
  bettiSum?: number;
}

export interface DeblurEnsembleResult {
  recovered: Float32Array; // pixel-wise weighted average across survivors
  width: number;
  height: number;
  channels: 1 | 3;
  candidates: ReadonlyArray<DeblurEnsembleCandidateReport>;
  fold: {
    rounds: number;
    survivors: number;
    rejected: number;
    totalWeight: number;
  };
  invisibleWatermark?: InvisibleWatermarkProvenance;
}

interface CandidateInternal {
  label: string;
  result: DeblurAnyImageResult;
  binetResidual: number;
  maxChannelScore: number;
  rejectedByCeiling: boolean;
}

interface FoldOutput {
  recovered: Float32Array;
  width: number;
  height: number;
  channels: 1 | 3;
  candidates: DeblurEnsembleCandidateReport[];
  totalWeight: number;
  survivors: number;
  rejected: number;
  invisibleWatermark?: InvisibleWatermarkProvenance;
}

/**
 * Handles the gnosis fork Race Fold Deblur workflow.
 */
export async function forkRaceFoldDeblur(
  args: DeblurEnsembleArgs,
): Promise<DeblurEnsembleResult> {
  const { input, configs, concurrency, ceiling: ceilingOption } = args;

  if (configs.length < 1) {
    throw new RangeError('forkRaceFoldDeblur: configs must be non-empty');
  }

  const ceiling = ceilingOption ?? TOWER_CEILINGS.hexon;
  if (!Number.isFinite(ceiling) || ceiling < 0) {
    throw new RangeError(
      `forkRaceFoldDeblur: ceiling must be >= 0, got ${ceiling}`,
    );
  }

  // FORK: build one task per config. Each task runs deblurAnyImage and
  // packages the result with derived ceiling/attractor metrics.
  const forks: ForkTask<CandidateInternal>[] = configs.map(
    (cfg, idx) => async () => {
      const result = await deblurAnyImage(input, {
        ...cfg.options,
        invisibleWatermark: { ...(cfg.options.invisibleWatermark ?? {}), enabled: false },
      });

      let maxChannelScore = 0;
      for (const ch of result.perChannel) {
        const s = ch.carrierBuley.score;
        if (s > maxChannelScore) maxChannelScore = s;
      }

      // Extract |binetResidual| from the phi-attractor result. When sigma
      // was explicit, no attractor ran — treat as best-case (residual = 0).
      let binetResidual = 0;
      const att = result.attractor;
      if (att !== undefined) {
        const best = att.candidates[att.bestCandidateIndex];
        if (best !== undefined) {
          binetResidual = Math.abs(best.binetResidual);
        }
      }

      const rejectedByCeiling = maxChannelScore > ceiling;

      return {
        label: cfg.label ?? `config-${idx}`,
        result,
        binetResidual,
        maxChannelScore,
        rejectedByCeiling,
      };
    },
  );

  // RACE: drop ceiling-violating candidates. Survivors are reported, then
  // we pull the FULL fork list (including rejected ones) out of the FOLD so
  // the candidates report covers every config — not just survivors.
  const allCandidates: CandidateInternal[] = [];
  // Side-channel diagnostics from the optional Betti sieve. Keyed by label
  // (labels are unique per fork — see `cfg.label ?? 'config-${idx}'`).
  // `bettiSumByLabel` is populated for every candidate the sieve evaluated;
  // `bettiRejectedByLabel` is the subset that failed isClosureAdmissible.
  const bettiSumByLabel = new Map<string, number>();
  const bettiRejectedByLabel = new Set<string>();

  const { output } = await forkRaceFold<CandidateInternal, FoldOutput>({
    forks,
    concurrency,
    abortOnReject: false,
    race: (results) => {
      // Side-effect capture: every fulfilled fork lands here, in fork-input
      // order. We keep the full list for the report and pass only ceiling-
      // honoring candidates downstream to FOLD.
      for (const r of results) allCandidates.push(r);
      return results.filter((r) => !r.rejectedByCeiling);
    },
    fold: (survivors): FoldOutput => {
      // Validate plane shape coherence — pixel-wise blend requires identical
      // dims across all survivors.
      const first = survivors[0]!;
      const { width, height, channels } = first.result;
      const planeLen = width * height * channels;
      for (let i = 1; i < survivors.length; i++) {
        const r = survivors[i]!.result;
        if (r.width !== width || r.height !== height || r.channels !== channels) {
          throw new RangeError(
            `forkRaceFoldDeblur: survivor ${i} dims (${r.width}x${r.height}x${r.channels}) mismatch survivor 0 (${width}x${height}x${channels})`,
          );
        }
        if (r.recovered.length !== planeLen) {
          throw new RangeError(
            `forkRaceFoldDeblur: survivor ${i} recovered.length=${r.recovered.length} != ${planeLen}`,
          );
        }
      }

      // Optional Betti sieve. Runs AFTER the ceiling filter and BEFORE the
      // rank/weight stage. For each survivor we extract the luminance plane
      // (channels===1 ⇒ pass-through; channels===3 ⇒ Rec.709 luma), compute
      // the (β0, β1, β2) signature, and drop survivors whose total exceeds
      // the closure ceiling. Rejected labels are stashed in
      // `bettiRejectedByLabel` so the post-fold full-report assembly can
      // emit them as rejectedByCeiling=true / rejectedByBettiClosure=true.
      let kept: CandidateInternal[] = [...survivors];
      if (args.bettiSieve) {
        const cfg = args.bettiSieve;
        const closureCeiling = cfg.ceiling ?? 10;
        const thresholdFraction = cfg.thresholdFraction ?? 0.5;
        const next: CandidateInternal[] = [];
        for (const s of survivors) {
          const plane = s.result.recovered;
          let lumi: Float32Array;
          if (channels === 1) {
            // Already single-channel — feed straight in.
            lumi = plane;
          } else {
            // RGB → Rec.709 luminance: 0.2126 R + 0.7152 G + 0.0722 B.
            // Plane is interleaved [R,G,B,R,G,B,...] of length w*h*3.
            const N = width * height;
            lumi = new Float32Array(N);
            for (let p = 0; p < N; p++) {
              const r = plane[p * 3]!;
              const g = plane[p * 3 + 1]!;
              const b = plane[p * 3 + 2]!;
              lumi[p] = 0.2126 * r + 0.7152 * g + 0.0722 * b;
            }
          }
          const sig = bettiSignature({
            plane: lumi,
            width,
            height,
            thresholdFraction,
          });
          bettiSumByLabel.set(s.label, sig.total);
          if (isClosureAdmissible(sig, closureCeiling)) {
            next.push(s);
          } else {
            bettiRejectedByLabel.add(s.label);
          }
        }
        if (next.length === 0) {
          throw new Error(
            'forkRaceFoldDeblur: no candidates survived Betti closure sieve',
          );
        }
        kept = next;
      }

      // Rank survivors by ascending |binetResidual| (lower = fewer rejections).
      const indexed = kept.map((s, i) => ({ s, i }));
      indexed.sort((a, b) => a.s.binetResidual - b.s.binetResidual);

      const R = kept.length;
      // rankByOriginalIndex[i] = rank of kept[i] (0 = best).
      const rankByOriginalIndex = new Array<number>(R);
      const weightByOriginalIndex = new Array<number>(R);
      for (let r = 0; r < indexed.length; r++) {
        const oi = indexed[r]!.i;
        rankByOriginalIndex[oi] = r;
        weightByOriginalIndex[oi] = buleyeanWeight(R, r);
      }

      // Optional rejection-weighted fold. When enabled, the per-survivor
      // weight is replaced by `rejectionWeightedFold(...).weights`, which
      // attenuates each survivor by its sigma-proximity to the rejected
      // candidates' sigmas (closer to a known-bad sigma ⇒ less trust).
      // The rejected pool here is the union of ceiling-rejected and (if the
      // Betti sieve ran) Betti-rejected candidates, drawn from
      // `allCandidates` so we have everyone's sigma in one place.
      if (args.rejectionWeights) {
        const alpha = args.rejectionWeights.alpha ?? 0.5;
        const survivingForFold: SurvivingCandidate[] = kept.map((s, i) => ({
          sigma: s.result.selectedSigma,
          rank: rankByOriginalIndex[i]!,
        }));
        const keptLabels = new Set(kept.map((s) => s.label));
        const rejectedForFold: RejectionPoint[] = [];
        for (const c of allCandidates) {
          if (keptLabels.has(c.label)) continue;
          rejectedForFold.push({
            sigma: c.result.selectedSigma,
            attractorDistance: c.binetResidual,
          });
        }
        const rwOut = rejectionWeightedFold({
          survivors: survivingForFold,
          rejected: rejectedForFold,
          alpha,
        });
        for (let i = 0; i < R; i++) {
          weightByOriginalIndex[i] = rwOut.weights[i]!;
        }
      }

      // Pixel-wise weighted average across (kept) survivors.
      const out = new Float32Array(planeLen);
      let totalW = 0;
      for (let i = 0; i < R; i++) totalW += weightByOriginalIndex[i]!;

      for (let i = 0; i < R; i++) {
        const w = weightByOriginalIndex[i]!;
        const plane = kept[i]!.result.recovered;
        for (let p = 0; p < planeLen; p++) {
          out[p]! += w * plane[p]!;
        }
      }
      const inv = 1 / totalW;
      for (let p = 0; p < planeLen; p++) out[p]! *= inv;

      // Build the per-survivor report rows. Rejected-by-ceiling and
      // rejected-by-Betti rows are appended after — see post-fold step.
      const survivorReports: DeblurEnsembleCandidateReport[] = kept.map(
        (s, i) => {
          const row: DeblurEnsembleCandidateReport = {
            label: s.label,
            selectedSigma: s.result.selectedSigma,
            methodUsed: s.result.methodUsed,
            binetResidual: s.binetResidual,
            maxChannelScore: s.maxChannelScore,
            rank: rankByOriginalIndex[i]!,
            buleyWeight: weightByOriginalIndex[i]!,
            rejectedByCeiling: false,
          };
          const bettiSum = bettiSumByLabel.get(s.label);
          if (bettiSum !== undefined) {
            row.bettiSum = bettiSum;
            row.rejectedByBettiClosure = false;
          }
          return row;
        },
      );

      const watermarked = applyInvisibleWatermarkToPlane({
        plane: out,
        width,
        height,
        channels,
        options: args.invisibleWatermark,
      });

      return {
        recovered: watermarked.plane,
        width,
        height,
        channels,
        candidates: survivorReports,
        totalWeight: totalW,
        survivors: R,
        rejected: 0, // set after this returns, from allCandidates
        invisibleWatermark: watermarked.provenance,
      };
    },
  }).catch((err: unknown) => {
    // Translate the generic FRF "no survivors after race" into the
    // domain-specific message the contract requires.
    if (err instanceof Error && /no survivors after race/.test(err.message)) {
      throw new Error(
        'forkRaceFoldDeblur: no candidates survived ceiling filter',
      );
    }
    throw err;
  });

  // Build the full candidate report — survivors (with their rank/weight) plus
  // ceiling-rejected entries with rank=-1, weight=0. We use allCandidates as
  // the canonical fork-input order, looking up survivor records by label.
  const survivorByLabel = new Map<string, DeblurEnsembleCandidateReport>();
  for (const sr of output.candidates) survivorByLabel.set(sr.label, sr);

  const fullReport: DeblurEnsembleCandidateReport[] = allCandidates.map(
    (c) => {
      const survivor = survivorByLabel.get(c.label);
      if (survivor !== undefined) return survivor;
      // Rejected — either by the carrier-ceiling RACE filter or by the
      // optional Betti closure sieve. `rejectedByCeiling` stays true as the
      // umbrella "removed" flag; `rejectedByBettiClosure` distinguishes
      // the Betti-specific subset when the sieve was enabled.
      const row: DeblurEnsembleCandidateReport = {
        label: c.label,
        selectedSigma: c.result.selectedSigma,
        methodUsed: c.result.methodUsed,
        binetResidual: c.binetResidual,
        maxChannelScore: c.maxChannelScore,
        rank: -1,
        buleyWeight: 0,
        rejectedByCeiling: true,
      };
      const bettiSum = bettiSumByLabel.get(c.label);
      if (bettiSum !== undefined) {
        row.bettiSum = bettiSum;
      }
      if (bettiRejectedByLabel.has(c.label)) {
        row.rejectedByBettiClosure = true;
      } else if (args.bettiSieve && bettiSum !== undefined) {
        row.rejectedByBettiClosure = false;
      }
      return row;
    },
  );
  const rejectedCount = fullReport.filter((r) => r.rejectedByCeiling).length;

  return {
    recovered: output.recovered,
    width: output.width,
    height: output.height,
    channels: output.channels,
    candidates: fullReport,
    fold: {
      rounds: output.survivors,
      survivors: output.survivors,
      rejected: rejectedCount,
      totalWeight: output.totalWeight,
    },
    invisibleWatermark: output.invisibleWatermark,
  };
}
