// @a0n/.ignored_gnosis/distributed-inference-host/src/forkRaceFoldStyleTransfer.ts
import { buleyeanWeight } from '@a0n/buleyean-kernel';
import { forkRaceFold, type ForkTask } from './frf';
import { compositeStyleTransferFoldScore } from './style-transfer-mesh-gate';
import { styleTransferAnyImage } from './style-transfer-any-image';
import type {
  StyleTransferAnyImageResult,
  StyleTransferCandidateReport,
  StyleTransferEnsembleArgs,
  StyleTransferEnsembleResult,
} from './style-transfer-types';

interface CandidateInternal {
  label: string;
  result: StyleTransferAnyImageResult;
  styleDistance: number;
  compositeFoldScore: number;
  rejectedByCeiling: boolean;
}

/**
 * Handles the gnosis fork Race Fold Style Transfer workflow.
 */
export async function forkRaceFoldStyleTransfer(
  args: StyleTransferEnsembleArgs,
): Promise<StyleTransferEnsembleResult> {
  if (args.configs.length < 1) {
    throw new RangeError('forkRaceFoldStyleTransfer: configs must be non-empty');
  }
  const ceiling = args.ceiling ?? Number.POSITIVE_INFINITY;
  if (
    ceiling !== Number.POSITIVE_INFINITY &&
    (!Number.isFinite(ceiling) || ceiling <= 0)
  ) {
    throw new RangeError(`forkRaceFoldStyleTransfer: ceiling must be > 0 or +Infinity, got ${ceiling}`);
  }

  const rankWeights = args.meshFoldRankWeights ?? {};
  const allCandidates: CandidateInternal[] = [];
  const forks: ForkTask<CandidateInternal>[] = args.configs.map((cfg, idx) => async () => {
    const result = await styleTransferAnyImage(
      args.content,
      args.style,
      cfg.options ?? {},
      undefined,
      args.topologyAlphaBackend,
      args.imagingWasm,
    );
    const compositeFoldScore = compositeStyleTransferFoldScore(
      result.styleDistance,
      result.meshFoldMetrics,
      rankWeights,
    );
    return {
      label: cfg.label ?? `style-config-${idx}`,
      styleDistance: result.styleDistance,
      compositeFoldScore,
      rejectedByCeiling: result.styleDistance > ceiling,
      result,
    };
  });

  const { output } = await forkRaceFold<CandidateInternal, StyleTransferEnsembleResult>({
    forks,
    concurrency: args.concurrency,
    abortOnReject: false,
    race: (items) => {
      for (const it of items) allCandidates.push(it);
      return items.filter((it) => !it.rejectedByCeiling);
    },
    fold: (survivors) => {
      const first = survivors[0]!;
      const width = first.result.width;
      const height = first.result.height;
      const channels = first.result.channels;
      const n = width * height * channels;

      const ranked = [...survivors].sort((a, b) => {
        const d = a.compositeFoldScore - b.compositeFoldScore;
        if (Math.abs(d) > 1e-12) return d;
        return a.styleDistance - b.styleDistance;
      });
      const rankByLabel = new Map<string, number>();
      for (let i = 0; i < ranked.length; i++) rankByLabel.set(ranked[i]!.label, i);

      let totalWeight = 0;
      const weights = new Map<string, number>();
      for (const c of survivors) {
        const rank = rankByLabel.get(c.label)!;
        const w = buleyeanWeight(survivors.length, rank);
        weights.set(c.label, w);
        totalWeight += w;
      }

      const out = new Float32Array(n);
      for (const c of survivors) {
        const w = weights.get(c.label)!;
        const plane = c.result.stylized;
        for (let i = 0; i < n; i++) out[i]! += w * plane[i]!;
      }
      const inv = 1 / totalWeight;
      for (let i = 0; i < n; i++) out[i]! *= inv;

      const survivorReports: StyleTransferCandidateReport[] = survivors.map((c) => ({
        label: c.label,
        methodUsed: c.result.methodUsed,
        rank: rankByLabel.get(c.label)!,
        buleyWeight: weights.get(c.label)!,
        rejectedByCeiling: false,
        styleDistance: c.styleDistance,
        compositeFoldScore: c.compositeFoldScore,
        meshFoldMetrics: c.result.meshFoldMetrics,
      }));

      return {
        stylized: out,
        width,
        height,
        channels,
        candidates: survivorReports,
        fold: {
          rounds: survivors.length,
          survivors: survivors.length,
          rejected: 0,
          totalWeight,
        },
      };
    },
  }).catch((err: unknown) => {
    if (err instanceof Error && /no survivors after race/.test(err.message)) {
      throw new Error('forkRaceFoldStyleTransfer: no candidates survived ceiling filter');
    }
    throw err;
  });

  const survivorMap = new Map(output.candidates.map((c) => [c.label, c]));
  const allReports: StyleTransferCandidateReport[] = allCandidates.map((c) => {
    const survivor = survivorMap.get(c.label);
    if (survivor !== undefined) return survivor;
    return {
      label: c.label,
      methodUsed: c.result.methodUsed,
      rank: -1,
      buleyWeight: 0,
      rejectedByCeiling: true,
      styleDistance: c.styleDistance,
      compositeFoldScore: c.compositeFoldScore,
      meshFoldMetrics: c.result.meshFoldMetrics,
    };
  });
  const rejected = allReports.filter((c) => c.rejectedByCeiling).length;
  return {
    ...output,
    candidates: allReports,
    fold: {
      ...output.fold,
      rejected,
    },
  };
}
