// @a0n/.ignored_gnosis/distributed-inference-host/src/frf-pool.ts
// TRUE parallel FRF via worker_threads.
// The WorkerFactory abstraction lets callers control how Worker instances
// are built — for tests, use `eval: true` with the function source as a
// string (no TS-loader dance); for production, use a file URL pointing to a
// pre-built JS worker; for dev, use a TS loader (tsx, @swc-node) via
// execArgv. The pool persists across calls so the dispatch path doesn't
// pay the worker spawn cost per task. Mirrors the role gnosis-frf plays in
// Rust: persistent worker pool, lock-free-ish dispatch, JS-side cost is
// the postMessage round-trip (~0.5ms vs gnosis-frf's 200ns).

import { Worker } from 'node:worker_threads';

export interface FRFPoolWorker {
  /** Underlying node:worker_threads Worker. */
  readonly worker: Worker;
  /** True when the worker is currently processing a task. */
  busy: boolean;
}

export type WorkerFactory = () => Worker;

export interface FRFPoolConfig {
  /** Number of workers. Must be >= 1. */
  size: number;
  /** How to build each worker. Caller chooses eval/file/loader. */
  workerFactory: WorkerFactory;
}

export interface PoolTask<I, O> {
  /** Arbitrary serialisable input passed to the worker via postMessage. */
  input: I;
  /** Optional list of ArrayBuffers to transfer (zero-copy). */
  transfer?: ArrayBuffer[];
  /** Phantom for output type inference. */
  readonly _outputBrand?: O;
}

interface PendingTask {
  task: PoolTask<unknown, unknown>;
  resolve: (value: unknown) => void;
  reject: (err: unknown) => void;
}

interface InFlightEntry {
  resolve: (value: unknown) => void;
  reject: (err: unknown) => void;
  workerEntry: FRFPoolWorker;
}

/** Inbound message format the WORKER must implement (caller's responsibility):
 *
 *   parentPort.on('message', async ({ taskId, input }) => {
 *     try {
 *       const output = await doWork(input);
 *       parentPort.postMessage({ taskId, output });
 *     } catch (err) {
 *       parentPort.postMessage({ taskId, error: String(err) });
 *     }
 *   });
 *
 * The pool tags each task with a numeric taskId; the worker echoes it back
 * so the pool can route the result to the right awaiter. */
export class FRFPool {
  private workers: FRFPoolWorker[] = [];
  private queue: PendingTask[] = [];
  private inFlightMap = new Map<number, InFlightEntry>();
  private nextTaskId = 1;
  private destroyed = false;
  private destroyPromise: Promise<void> | null = null;

  constructor(config: FRFPoolConfig) {
    if (typeof config !== 'object' || config === null) {
      throw new TypeError('FRFPool: config must be an object');
    }
    if (typeof config.size !== 'number' || !Number.isInteger(config.size) || config.size < 1) {
      throw new RangeError('FRFPool: size must be an integer >= 1');
    }
    if (typeof config.workerFactory !== 'function') {
      throw new TypeError('FRFPool: workerFactory must be a function');
    }

    for (let i = 0; i < config.size; i++) {
      const worker = config.workerFactory();
      const entry: FRFPoolWorker = { worker, busy: false };
      this.attachHandlers(entry);
      this.workers.push(entry);
    }
  }

  private attachHandlers(entry: FRFPoolWorker): void {
    entry.worker.on('message', (msg: { taskId: number; output?: unknown; error?: unknown }) => {
      if (!msg || typeof msg.taskId !== 'number') return;
      const pending = this.inFlightMap.get(msg.taskId);
      if (!pending) return;
      this.inFlightMap.delete(msg.taskId);
      entry.busy = false;

      if ('error' in msg && msg.error !== undefined) {
        pending.reject(new Error(String(msg.error)));
      } else {
        pending.resolve(msg.output);
      }

      // Drain queue: try to dispatch a queued task to this freed worker.
      this.drain();
    });

    entry.worker.on('error', (err) => {
      // Reject every in-flight task on this worker.
      for (const [taskId, inFlight] of this.inFlightMap.entries()) {
        if (inFlight.workerEntry === entry) {
          this.inFlightMap.delete(taskId);
          inFlight.reject(err);
        }
      }
      // Remove this worker from the pool.
      const idx = this.workers.indexOf(entry);
      if (idx >= 0) this.workers.splice(idx, 1);
      entry.busy = false;
    });
  }

  private drain(): void {
    while (this.queue.length > 0) {
      const idleIdx = this.workers.findIndex((w) => !w.busy);
      if (idleIdx < 0) return;
      const pending = this.queue.shift()!;
      this.dispatch(this.workers[idleIdx], pending.task, pending.resolve, pending.reject);
    }
  }

  private dispatch(
    entry: FRFPoolWorker,
    task: PoolTask<unknown, unknown>,
    resolve: (value: unknown) => void,
    reject: (err: unknown) => void,
  ): void {
    const taskId = this.nextTaskId++;
    entry.busy = true;
    this.inFlightMap.set(taskId, { resolve, reject, workerEntry: entry });

    try {
      const message = { taskId, input: task.input };
      if (task.transfer && task.transfer.length > 0) {
        entry.worker.postMessage(message, task.transfer);
      } else {
        entry.worker.postMessage(message);
      }
    } catch (err) {
      this.inFlightMap.delete(taskId);
      entry.busy = false;
      reject(err);
    }
  }

  /** Run a single task on any idle worker. Awaits a free slot if all are busy. */
  run<I, O>(task: PoolTask<I, O>): Promise<O> {
    if (this.destroyed) {
      return Promise.reject(new Error('FRFPool: pool destroyed'));
    }
    if (typeof task !== 'object' || task === null) {
      return Promise.reject(new TypeError('FRFPool.run: task must be an object'));
    }

    return new Promise<O>((resolve, reject) => {
      const idleIdx = this.workers.findIndex((w) => !w.busy);
      const wrappedResolve = (v: unknown) => resolve(v as O);
      if (idleIdx >= 0) {
        this.dispatch(
          this.workers[idleIdx],
          task as PoolTask<unknown, unknown>,
          wrappedResolve,
          reject,
        );
      } else {
        this.queue.push({
          task: task as PoolTask<unknown, unknown>,
          resolve: wrappedResolve,
          reject,
        });
      }
    });
  }

  /** FORK an array of tasks; RACE filter; FOLD survivors. The worker-thread
   *  analog of frf.forkRaceFold — tasks actually run in parallel. */
  async forkRaceFold<I, O, U>(args: {
    inputs: ReadonlyArray<I>;
    transferFor?: (input: I) => ArrayBuffer[] | undefined;
    race?: (results: ReadonlyArray<O>) => ReadonlyArray<O>;
    fold: (survivors: ReadonlyArray<O>) => U;
  }): Promise<U> {
    if (this.destroyed) {
      throw new Error('FRFPool: pool destroyed');
    }
    if (!args || !Array.isArray(args.inputs)) {
      throw new TypeError('FRFPool.forkRaceFold: inputs must be an array');
    }
    if (args.inputs.length === 0) {
      throw new RangeError('FRFPool.forkRaceFold: inputs must be non-empty');
    }
    if (typeof args.fold !== 'function') {
      throw new TypeError('FRFPool.forkRaceFold: fold must be a function');
    }

    const { inputs, transferFor, race, fold } = args;

    // Preserve fork-input order: launch in order, await Promise.all,
    // results array index === input index.
    const promises = inputs.map((input) => {
      const transfer = transferFor ? transferFor(input) : undefined;
      const task: PoolTask<I, O> = transfer ? { input, transfer } : { input };
      return this.run<I, O>(task);
    });

    const results = await Promise.all(promises);
    const survivors = race ? race(results) : results;
    return fold(survivors);
  }

  /** Number of workers currently in the pool. */
  size(): number {
    return this.workers.length;
  }

  /** Number of tasks currently in flight. */
  inFlight(): number {
    return this.inFlightMap.size;
  }

  /** Terminate all workers; pool becomes unusable. Idempotent. */
  destroy(): Promise<void> {
    if (this.destroyPromise) return this.destroyPromise;
    this.destroyed = true;

    this.destroyPromise = (async () => {
      const workers = this.workers.slice();
      // Best-effort shutdown signal — workers can ignore.
      for (const entry of workers) {
        try {
          entry.worker.postMessage({ type: 'shutdown' });
        } catch {
          // Worker may already be dead; ignore.
        }
      }
      // Hard terminate.
      await Promise.all(
        workers.map(async (entry) => {
          try {
            await entry.worker.terminate();
          } catch {
            // Already terminated; ignore.
          }
        }),
      );
      this.workers = [];

      // Reject any queued tasks.
      for (const pending of this.queue) {
        pending.reject(new Error('FRFPool: pool destroyed'));
      }
      this.queue = [];

      // Reject any in-flight tasks that didn't get a response.
      for (const [, inFlight] of this.inFlightMap.entries()) {
        inFlight.reject(new Error('FRFPool: pool destroyed'));
      }
      this.inFlightMap.clear();
    })();

    return this.destroyPromise;
  }
}

/** Convenience: run a one-shot FRF over an array of inputs, building the pool,
 *  running the FRF, and tearing it down. Equivalent to:
 *    const p = new FRFPool({ size, workerFactory });
 *    try { return await p.forkRaceFold({...}); } finally { await p.destroy(); }
 *  Prefer FRFPool directly when you have multiple FRF calls — the pool
 *  amortises the worker-spawn cost across calls. */
export async function withFRFPool<I, O, U>(args: {
  size: number;
  workerFactory: WorkerFactory;
  inputs: ReadonlyArray<I>;
  transferFor?: (input: I) => ArrayBuffer[] | undefined;
  race?: (results: ReadonlyArray<O>) => ReadonlyArray<O>;
  fold: (survivors: ReadonlyArray<O>) => U;
}): Promise<U> {
  const pool = new FRFPool({ size: args.size, workerFactory: args.workerFactory });
  try {
    return await pool.forkRaceFold<I, O, U>({
      inputs: args.inputs,
      transferFor: args.transferFor,
      race: args.race,
      fold: args.fold,
    });
  } finally {
    await pool.destroy();
  }
}
