// @a0n/.ignored_gnosis/distributed-inference-host/src/host-shim.ts
/**
 * HostShim — the orchestration layer between async byte-source fetches and
 * the synchronous WASM-bindgen calls produced by the distributed-inference
 * Rust crate.
 *
 * Lifecycle:
 *   1. open(source, wasmModule, opts)
 *        - fetches the knot header
 *        - constructs the WasmHostBackend with the metadata JSON
 *        - constructs the WasmLlamaPipeline against the backend
 *        - loads the resident tensors (token_embd, output_norm, output)
 *   2. prefetchLayer(L) repeatedly, OR generate(prompt, max) for the simple path
 *   3. dispose() to release WASM-side memory
 */

import type {
  HostShimOptions,
  KnotByteSource,
  TensorInfo,
  WasmHostBackendLike,
  WasmLlamaPipelineLike,
  WasmModule,
} from './types';
import { readKnotHeader, type KnotHeader } from './knot-header';
import { LayerLRU } from './lru';
import {
  PER_LAYER_TENSOR_SUFFIXES,
  RESIDENT_TENSOR_NAMES,
  tensorNamesForLayer,
} from './tensor-names';

const DEFAULT_HEADER_FETCH_BYTES = 4 * 1024 * 1024; // 4 MiB
const DEFAULT_LRU_CAPACITY = 4;
const DEFAULT_RESONANCE_THRESHOLD = 0.0;

export class HostShim {
  private constructor(
    private readonly source: KnotByteSource,
    private readonly wasmModule: WasmModule,
    private readonly backend: WasmHostBackendLike,
    private readonly pipeline: WasmLlamaPipelineLike,
    private readonly header: KnotHeader,
    private readonly lru: LayerLRU,
    private readonly residentNames: string[],
    private disposed: boolean,
  ) {}

  static async open(
    source: KnotByteSource,
    wasmModule: WasmModule,
    options: HostShimOptions = {},
  ): Promise<HostShim> {
    const headerFetchBytes = options.headerFetchBytes ?? DEFAULT_HEADER_FETCH_BYTES;
    const lruCapacity = options.lruCapacity ?? DEFAULT_LRU_CAPACITY;
    const draftMode = options.draftMode ?? false;
    const resonanceThreshold = options.resonanceThreshold ?? DEFAULT_RESONANCE_THRESHOLD;

    const header = await readKnotHeader(source, headerFetchBytes);

    // Construct the backend. Prefer the WASM-side parsers when present so the
    // backend sees the same byte-for-byte JSON the Rust crate would have emitted.
    const tensorsJson = wasmModule.parse_tensors_json
      ? wasmModule.parse_tensors_json(header.metadataJson)
      : JSON.stringify(header.metadata.tensors);
    const configJson = wasmModule.parse_config_json
      ? wasmModule.parse_config_json(header.metadataJson)
      : header.metadata.config != null
        ? JSON.stringify(header.metadata.config)
        : null;

    const backend = new wasmModule.WasmHostBackend(tensorsJson);
    const pipeline = wasmModule.WasmLlamaPipeline.from_backend(
      backend,
      configJson,
      header.payloadOffset,
      draftMode,
      resonanceThreshold,
    );

    const lru = new LayerLRU(lruCapacity);

    const shim = new HostShim(
      source,
      wasmModule,
      backend,
      pipeline,
      header,
      lru,
      [],
      false,
    );

    // Resident tensors loaded once. `output_weight` is optional — skip if the
    // metadata didn't list it (tied embeddings).
    const residentLoaded: string[] = [];
    for (const name of RESIDENT_TENSOR_NAMES) {
      const info = backend.tensor_info(name);
      if (info == null) continue;
      const bytes = await shim.fetchTensorBytes(name, info);
      backend.set_tensor(name, bytes);
      residentLoaded.push(name);
    }
    // Mutate residentNames in place since shim was constructed with [].
    residentLoaded.forEach((name) => shim.residentNames.push(name));

    return shim;
  }

  /**
   * Fetch + populate every per-layer tensor for layer `layerIdx`. All nine
   * range gets fire in parallel; the function only resolves once every
   * `set_tensor` call has been made on the backend.
   */
  async prefetchLayer(layerIdx: number): Promise<void> {
    this.assertOpen();

    const names = tensorNamesForLayer(layerIdx);

    // Resolve infos before issuing fetches so a missing tensor surfaces
    // before we burn bytes from the byte source.
    const infos: { name: string; info: TensorInfo }[] = names.map((name) => {
      const info = this.backend.tensor_info(name);
      if (info == null) {
        throw new Error(
          `prefetchLayer(${layerIdx}): backend has no tensor_info for "${name}". ` +
            `Knot is likely missing this tensor or the layer index is out of range.`,
        );
      }
      return { name, info };
    });

    // Fan out range fetches; await all before mutating the backend so the LRU
    // bookkeeping reflects a fully-populated layer.
    const fetched = await Promise.all(
      infos.map(({ name, info }) => this.fetchTensorBytes(name, info)),
    );

    for (let i = 0; i < infos.length; i += 1) {
      const desc = infos[i];
      const bytes = fetched[i];
      if (!desc || !bytes) continue;
      this.backend.set_tensor(desc.name, bytes);
    }

    // Now bump LRU. If the touch evicts a previous layer, fire `evict` for
    // its nine names. Residents never enter the LRU and are never evicted.
    const victim = this.lru.recordTouch(layerIdx);
    if (victim != null && victim !== layerIdx) {
      const victimNames = tensorNamesForLayer(victim);
      this.backend.evict(victimNames);
    }
  }

  /**
   * Convenience generation path. Pre-loads ALL transformer layers, then runs
   * the WASM pipeline. This is the proof-of-deploy path; larger models will
   * want to interleave per-step prefetch via prefetchLayer() and a future
   * pipeline.step() API.
   *
   * Note: when `lruCapacity < numLayers`, eviction will occur during the
   * preload sweep. The shim still issues every prefetch in declared order so
   * later layers are guaranteed to be resident at generate() time.
   */
  async generate(promptTokens: Uint32Array, maxNewTokens: number): Promise<Uint32Array> {
    this.assertOpen();
    const numLayers = this.detectLayerCount();
    if (numLayers <= 0) {
      throw new Error(
        'HostShim.generate: could not detect number of layers from knot metadata.',
      );
    }
    for (let l = 0; l < numLayers; l += 1) {
      // Sequential await on purpose: preserves LRU order under tight capacity
      // budgets. Within each call, the nine tensor fetches still parallelize.
      // eslint-disable-next-line no-await-in-loop
      await this.prefetchLayer(l);
    }
    return this.pipeline.generate(promptTokens, maxNewTokens);
  }

  /** Current pipeline position (KV cache fill). */
  getPos(): number {
    this.assertOpen();
    return this.pipeline.get_pos();
  }

  /** Layers currently resident in the LRU, ordered LRU -> MRU. */
  residentLayers(): number[] {
    return this.lru.residentLayers();
  }

  /** Names of resident (never-evicted) tensors loaded at open. */
  residentTensorNames(): string[] {
    return this.residentNames.slice();
  }

  /** Parsed knot header (read-only). */
  knotHeader(): KnotHeader {
    return this.header;
  }

  /**
   * Drop references to the WASM-allocated objects. After this call, all
   * methods throw. Idempotent.
   */
  dispose(): void {
    if (this.disposed) return;
    this.disposed = true;
    // wasm-bindgen produced classes typically expose a `.free()`. Call it if
    // present, but tolerate mocks that don't.
    callIfFn(this.pipeline as unknown, 'free');
    callIfFn(this.backend as unknown, 'free');
    this.lru.clear();
  }

  // ---- internals ----

  /**
   * Fetch the exact byte range for a single tensor and return it sized to
   * `info.length`.
   *
   * Why this function refuses to zero-pad short reads
   * --------------------------------------------------
   * A tensor read that comes back short is *not* a recoverable warning
   * condition — it is silent model corruption waiting to be deployed. The
   * distributed-inference pipeline treats every byte returned here as the
   * authoritative weight buffer for a specific named tensor (token_embd,
   * blk_N_attn_q_weight, output_norm, ...). Those bytes flow directly into
   * the WASM backend via `set_tensor`, which interprets them as packed
   * F16 / Q4_K / F32 entries with no further validation.
   *
   * If we zero-pad a short read:
   *   - The trailing region of that tensor becomes all-zero entries.
   *   - For embeddings this maps high-vocab tokens to the zero vector,
   *     which silently collapses meaning while still producing a valid-
   *     looking probability distribution downstream.
   *   - For attention Q/K/V projections, zeroed rows make late-position
   *     channels effectively dead while early-position channels still fire,
   *     producing fluent-but-wrong continuations.
   *   - For FFN up/gate/down weights, partial zero-fill skews the
   *     activation magnitude in a direction the residual stream learns to
   *     compensate for over time — meaning loss looks roughly normal in
   *     short evals and only blows up on long-context, niche-vocab, or
   *     reasoning-heavy prompts.
   *
   * The combined failure mode is "coherent-looking gibberish": tokens are
   * grammatically plausible, perplexity on common prefixes looks fine,
   * but the model has effectively been lobotomized in places no
   * smoke-test will catch. We have spent multi-day debugging sessions
   * chasing exactly this signature elsewhere in the stack; refusing the
   * partial read here surfaces the bug in the *first* request after a
   * deploy, not weeks later as a quality regression nobody can root-
   * cause.
   *
   * Over-reads are still tolerated: byte-source caches and aligned-block
   * fetchers may legitimately return a few extra bytes past the requested
   * range. We trim those to `info.length` because the extra bytes are
   * never tensor data — they belong to the next tensor, padding, or an
   * unrelated cache page boundary.
   *
   * @throws Error if the source returns fewer than `info.length` bytes.
   *   The error names the tensor, both byte counts, and explains why we
   *   refuse to pad. Callers should not catch and retry with padding;
   *   the deploy is broken.
   */
  private async fetchTensorBytes(name: string, info: TensorInfo): Promise<Uint8Array> {
    const start = this.header.payloadOffset + info.offset;
    const end = start + info.length - 1; // inclusive
    let lastError: unknown = null;
    for (let attempt = 0; attempt < 2; attempt += 1) {
      try {
        const bytes = await this.source.rangeGet(start, end);
        if (bytes.length < info.length) {
          throw new Error(
            `fetchTensorBytes(${name}): source returned ${bytes.length} bytes, expected ${info.length}. ` +
              `Refusing to zero-pad — partial reads silently corrupt downstream inference.`,
          );
        }
        // Trim if the source over-read (cache granularity can do this).
        return bytes.length === info.length ? bytes : bytes.subarray(0, info.length);
      } catch (err) {
        lastError = err;
        if (attempt === 0) {
          continue;
        }
        throw err;
      }
    }
    throw lastError instanceof Error
      ? lastError
      : new Error(`fetchTensorBytes(${name}): failed after retry`);
  }

  private detectLayerCount(): number {
    // 1. Prefer config.num_layers if present in the knot metadata.
    const cfg = this.header.metadata.config;
    if (cfg && typeof cfg === 'object') {
      const candidates = ['num_layers', 'n_layer', 'num_hidden_layers'];
      for (const key of candidates) {
        const v = (cfg as Record<string, unknown>)[key];
        if (typeof v === 'number' && Number.isInteger(v) && v > 0) {
          return v;
        }
      }
    }
    // 2. Fall back to scanning tensor names: highest blk_N_attn_norm_weight + 1.
    let maxIdx = -1;
    const re = /^blk_(\d+)_/;
    for (const name of Object.keys(this.header.metadata.tensors)) {
      const m = re.exec(name);
      if (!m || !m[1]) continue;
      const idx = Number.parseInt(m[1], 10);
      if (Number.isFinite(idx) && idx > maxIdx) {
        maxIdx = idx;
      }
    }
    return maxIdx + 1;
  }

  private assertOpen(): void {
    if (this.disposed) {
      throw new Error('HostShim: instance has been disposed.');
    }
  }
}

function callIfFn(obj: unknown, key: string): void {
  if (obj != null && typeof obj === 'object') {
    const fn = (obj as Record<string, unknown>)[key];
    if (typeof fn === 'function') {
      try {
        (fn as () => void).call(obj);
      } catch {
        // disposal must never throw
      }
    }
  }
}

// Re-export tensor-name helpers so the shim's main module is one-stop.
export { PER_LAYER_TENSOR_SUFFIXES, RESIDENT_TENSOR_NAMES, tensorNamesForLayer };
