// @a0n/.ignored_gnosis/distributed-inference-host/src/image-bilateral-frf.ts
// Row-tiled bilateral filter via the frf primitive. Output is bit-exact
// equivalent to image-bilateral.ts on 1-tile inputs; for K>=2 tiles each
// FORK computes a horizontal stripe of (interior + 2*kernelRadius overlap)
// rows, then FOLD concatenates the interior rows of each stripe into the
// final output. Async-shaped via Promise.all (CPU-bound; single-thread
// Node serialises microtasks but the structure is FRF-canonical and
// transparently parallelises behind worker_threads / WASM SIMD when wired
// up).

import { bilateralFilter } from './image-bilateral';
import { forkRaceFold, type ForkTask } from './frf';

export interface BilateralFRFArgs {
  plane: Float32Array;
  width: number;
  height: number;
  sigmaSpatial?: number;     // default 2.0
  sigmaIntensity?: number;   // default 0.05
  kernelRadius?: number;     // default ceil(3 * sigmaSpatial)
  /** Number of horizontal stripes. Default min(8, height). Must be 1..height. */
  tiles?: number;
  /** Concurrency cap for the underlying forkRaceFold. Default = tiles. */
  concurrency?: number;
}

interface TileResult {
  index: number;
  yStart: number;
  yEnd: number;
  central: Float32Array; // length = (yEnd - yStart) * width
}

/**
 * Handles the gnosis bilateral Filter FRF workflow.
 */
export async function bilateralFilterFRF(
  args: BilateralFRFArgs,
): Promise<Float32Array> {
  const { plane, width, height } = args;
  const sigmaSpatial = args.sigmaSpatial ?? 2.0;
  const sigmaIntensity = args.sigmaIntensity ?? 0.05;
  const kernelRadius = args.kernelRadius ?? Math.ceil(3 * sigmaSpatial);
  const tiles = args.tiles ?? Math.min(8, height);

  if (!(width > 0)) {
    throw new RangeError(
      `bilateralFilterFRF: width must be > 0 (got ${width})`,
    );
  }
  if (!(height > 0)) {
    throw new RangeError(
      `bilateralFilterFRF: height must be > 0 (got ${height})`,
    );
  }
  if (plane.length !== width * height) {
    throw new RangeError(
      `bilateralFilterFRF: plane.length (${plane.length}) !== width*height (${width * height})`,
    );
  }
  if (!(sigmaSpatial > 0)) {
    throw new RangeError(
      `bilateralFilterFRF: sigmaSpatial must be > 0 (got ${sigmaSpatial})`,
    );
  }
  if (!(sigmaIntensity > 0)) {
    throw new RangeError(
      `bilateralFilterFRF: sigmaIntensity must be > 0 (got ${sigmaIntensity})`,
    );
  }
  if (!(kernelRadius >= 0) || !Number.isFinite(kernelRadius)) {
    throw new RangeError(
      `bilateralFilterFRF: kernelRadius must be >= 0 (got ${kernelRadius})`,
    );
  }
  if (
    !Number.isFinite(tiles) ||
    !Number.isInteger(tiles) ||
    tiles < 1 ||
    tiles > height
  ) {
    throw new RangeError(
      `bilateralFilterFRF: tiles must be an integer in [1, height=${height}] (got ${tiles})`,
    );
  }
  if (
    args.concurrency !== undefined &&
    (!Number.isFinite(args.concurrency) || args.concurrency < 1)
  ) {
    throw new RangeError(
      `bilateralFilterFRF: concurrency must be >= 1 (got ${args.concurrency})`,
    );
  }

  const r = kernelRadius | 0;

  // Fast path: single tile (or no overlap needed). Just call bilateralFilter
  // and return a fresh Float32Array (bilateralFilter already allocates fresh).
  if (tiles === 1 || r === 0) {
    const full = bilateralFilter({
      plane,
      width,
      height,
      sigmaSpatial,
      sigmaIntensity,
      kernelRadius: r,
    });
    if (tiles === 1) return full;
    // r === 0 with tiles > 1: still return the full result (no overlap math
    // needed — every pixel only depends on itself). Keep the FRF shape
    // implicit; the answer is identical regardless of partitioning.
    return full;
  }

  const stripeRows = Math.ceil(height / tiles);

  const buildTask = (t: number): ForkTask<TileResult> => {
    const yStart = t * stripeRows;
    const yEnd = Math.min(height, yStart + stripeRows);
    const yPadStart = Math.max(0, yStart - r);
    const yPadEnd = Math.min(height, yEnd + r);
    const topPad = yStart - yPadStart;
    const subHeight = yPadEnd - yPadStart;
    const subLength = subHeight * width;

    return async () => {
      // Slice the input plane into a contiguous sub-plane covering
      // [yPadStart, yPadEnd) rows.
      const sub = new Float32Array(subLength);
      sub.set(
        plane.subarray(yPadStart * width, yPadEnd * width),
        0,
      );

      const filtered = bilateralFilter({
        plane: sub,
        width,
        height: subHeight,
        sigmaSpatial,
        sigmaIntensity,
        kernelRadius: r,
      });

      // Extract central rows: rows [topPad, topPad + (yEnd - yStart)).
      const centralRows = yEnd - yStart;
      const central = new Float32Array(centralRows * width);
      central.set(
        filtered.subarray(topPad * width, (topPad + centralRows) * width),
        0,
      );

      return { index: t, yStart, yEnd, central };
    };
  };

  const forks: Array<ForkTask<TileResult>> = [];
  for (let t = 0; t < tiles; t++) {
    // Skip empty tiles (can happen when stripeRows*tiles > height; here only
    // possible if tiles > height which we already rejected, but keep defensive).
    const yStart = t * stripeRows;
    if (yStart >= height) break;
    forks.push(buildTask(t));
  }

  const concurrency = args.concurrency ?? forks.length;

  const { output } = await forkRaceFold<TileResult, Float32Array>({
    forks,
    fold: (survivors) => {
      const out = new Float32Array(width * height);
      // Survivors arrive in fork-input order (forkRaceFold preserves order).
      for (const tile of survivors) {
        out.set(tile.central, tile.yStart * width);
      }
      return out;
    },
    concurrency,
  });

  return output;
}
