// @a0n/.ignored_gnosis/distributed-inference-host/src/image-carrier-pipeline.ts
/**
 * Image carrier pipeline — image adapter for the generic carrier medium.
 *
 * This module does not deblur pixels yet. It prepares the carrier witness data
 * a later deblur sampler needs: patch luminance, source carrier position,
 * observer-relative evolution/resolution readings, and the fiber/copy-index
 * losses caused by projecting a high-band image carrier down to a smaller one.
 */

import {
  CarrierMediumPipeline,
  type CarrierMediumBandwidthSummary,
  type CarrierMediumCellInput,
  type CarrierMediumReading,
} from './carrier-medium';
import {
  BANDWIDTH_HEXON,
  BANDWIDTH_TRIHEXENNEON,
  BANDWIDTH_TRITON,
} from './pleromatic-bandwidth';

export {
  analyzeCarrierMedium,
  carrierFiberForPosition,
  classifyCarrierStep,
  projectCarrierPosition,
  quantizeUnitToCarrierStep,
} from './carrier-medium';
export type {
  CarrierFiber,
  CarrierMediumReading as ImageCarrierReading,
  CarrierMediumReadingKind as ImageCarrierReadingKind,
  CarrierMediumBandwidthSummary as ImageCarrierBandwidthSummary,
} from './carrier-medium';

export type ImageSampleRange = 'unit' | 'byte';

export interface ImageCarrierInput {
  readonly data: ArrayLike<number>;
  readonly width: number;
  readonly height: number;
  readonly channels?: number;
  readonly sampleRange?: ImageSampleRange;
}

export interface ImageCarrierPipelineConfig {
  readonly carrierBandwidth?: number;
  /** @deprecated Use `carrierBandwidth`; retained for the first image adapter cut. */
  readonly highBandwidth?: number;
  readonly observerBandwidths?: readonly number[];
  readonly patchSize?: number;
}

export interface ImageCarrierPatch {
  readonly patchX: number;
  readonly patchY: number;
  readonly pixelX: number;
  readonly pixelY: number;
  readonly width: number;
  readonly height: number;
  readonly meanLuminance: number;
  readonly stepOrigin: number;
  readonly nextPosition: number;
  readonly readings: readonly CarrierMediumReading[];
}

export interface ImageCarrierAnalysis {
  readonly width: number;
  readonly height: number;
  readonly channels: number;
  readonly patchSize: number;
  readonly patchColumns: number;
  readonly patchRows: number;
  readonly carrierBandwidth: number;
  /** @deprecated Use `carrierBandwidth`; retained for consumers of the first cut. */
  readonly highBandwidth: number;
  readonly observerBandwidths: readonly number[];
  readonly patches: readonly ImageCarrierPatch[];
  readonly summaries: readonly CarrierMediumBandwidthSummary[];
  readonly buleCostInvariantAcrossBandwidths: boolean;
}

interface ImagePatchCarrierMetadata {
  readonly patchX: number;
  readonly patchY: number;
  readonly pixelX: number;
  readonly pixelY: number;
  readonly width: number;
  readonly height: number;
  readonly meanLuminance: number;
}

export class ImageCarrierPipeline {
  private readonly carrierBandwidth: number;
  private readonly observerBandwidths: readonly number[];
  private readonly patchSize: number;
  private readonly carrierMedium: CarrierMediumPipeline;

  constructor(config: ImageCarrierPipelineConfig = {}) {
    const carrierBandwidth = resolveCarrierBandwidth(config);
    const observerBandwidths = config.observerBandwidths;
    const patchSize = config.patchSize ?? 1;
    assertPositiveInteger(patchSize, 'patchSize');

    this.carrierBandwidth = carrierBandwidth;
    this.observerBandwidths = Array.from(
      observerBandwidths ?? [
        BANDWIDTH_TRITON,
        BANDWIDTH_HEXON,
        BANDWIDTH_TRIHEXENNEON,
      ],
    );
    this.patchSize = patchSize;
    this.carrierMedium = new CarrierMediumPipeline({
      sourceBandwidth: carrierBandwidth,
      observerBandwidths: this.observerBandwidths,
    });
  }

  analyze(input: ImageCarrierInput): ImageCarrierAnalysis {
    const width = assertPositiveInteger(input.width, 'width');
    const height = assertPositiveInteger(input.height, 'height');
    const channels = assertPositiveInteger(input.channels ?? 1, 'channels');
    const expectedSamples = width * height * channels;
    if (input.data.length !== expectedSamples) {
      throw new Error(
        `ImageCarrierPipeline.analyze: data length ${input.data.length} ` +
          `does not match width * height * channels (${expectedSamples})`,
      );
    }

    const patchColumns = Math.ceil(width / this.patchSize);
    const patchRows = Math.ceil(height / this.patchSize);
    const cells: CarrierMediumCellInput<ImagePatchCarrierMetadata>[] = [];
    for (let patchY = 0; patchY < patchRows; patchY += 1) {
      for (let patchX = 0; patchX < patchColumns; patchX += 1) {
        const pixelX = patchX * this.patchSize;
        const pixelY = patchY * this.patchSize;
        const patchWidth = Math.min(this.patchSize, width - pixelX);
        const patchHeight = Math.min(this.patchSize, height - pixelY);
        const meanLuminance = patchMeanLuminance(
          input,
          channels,
          pixelX,
          pixelY,
          patchWidth,
          patchHeight,
        );
        cells.push({
          unitValue: meanLuminance,
          metadata: {
            patchX,
            patchY,
            pixelX,
            pixelY,
            width: patchWidth,
            height: patchHeight,
            meanLuminance,
          },
        });
      }
    }

    const carrierAnalysis = this.carrierMedium.analyze({ cells });
    const patches = carrierAnalysis.cells.map((cell) => {
      const metadata = cell.metadata;
      if (!metadata) {
        throw new Error('ImageCarrierPipeline: missing image patch metadata');
      }
      return {
        ...metadata,
        stepOrigin: cell.stepOrigin,
        nextPosition: cell.nextPosition,
        readings: cell.readings,
      };
    });

    return {
      width,
      height,
      channels,
      patchSize: this.patchSize,
      patchColumns,
      patchRows,
      carrierBandwidth: carrierAnalysis.sourceBandwidth,
      highBandwidth: carrierAnalysis.sourceBandwidth,
      observerBandwidths: carrierAnalysis.observerBandwidths,
      patches,
      summaries: carrierAnalysis.summaries,
      buleCostInvariantAcrossBandwidths:
        carrierAnalysis.buleCostInvariantAcrossBandwidths,
    };
  }
}

function resolveCarrierBandwidth(config: ImageCarrierPipelineConfig): number {
  const carrierBandwidth = config.carrierBandwidth ?? config.highBandwidth ??
    BANDWIDTH_TRIHEXENNEON;
  if (
    config.carrierBandwidth !== undefined &&
    config.highBandwidth !== undefined &&
    config.carrierBandwidth !== config.highBandwidth
  ) {
    throw new Error(
      'ImageCarrierPipeline: carrierBandwidth and highBandwidth must match when both are provided',
    );
  }
  return assertPositiveInteger(carrierBandwidth, 'carrierBandwidth');
}

function patchMeanLuminance(
  input: ImageCarrierInput,
  channels: number,
  pixelX: number,
  pixelY: number,
  patchWidth: number,
  patchHeight: number,
): number {
  let sum = 0;
  let count = 0;
  for (let y = pixelY; y < pixelY + patchHeight; y += 1) {
    for (let x = pixelX; x < pixelX + patchWidth; x += 1) {
      const offset = (y * input.width + x) * channels;
      sum += pixelLuminance(input.data, offset, channels, input.sampleRange);
      count += 1;
    }
  }
  return count === 0 ? 0 : sum / count;
}

function pixelLuminance(
  data: ArrayLike<number>,
  offset: number,
  channels: number,
  sampleRange: ImageSampleRange | undefined,
): number {
  if (channels >= 3) {
    const r = sampleToUnit(data[offset] ?? 0, sampleRange);
    const g = sampleToUnit(data[offset + 1] ?? 0, sampleRange);
    const b = sampleToUnit(data[offset + 2] ?? 0, sampleRange);
    return clamp01(0.2126 * r + 0.7152 * g + 0.0722 * b);
  }
  return sampleToUnit(data[offset] ?? 0, sampleRange);
}

function sampleToUnit(value: number, sampleRange: ImageSampleRange | undefined): number {
  if (!Number.isFinite(value)) return 0;
  if (sampleRange === 'byte') return clamp01(value / 255);
  if (sampleRange === 'unit') return clamp01(value);
  return clamp01(value > 1 ? value / 255 : value);
}

function assertPositiveInteger(value: number, label: string): number {
  if (!Number.isInteger(value) || value <= 0) {
    throw new RangeError(`${label} must be a positive integer`);
  }
  return value;
}

function clamp01(value: number): number {
  if (value <= 0) return 0;
  if (value >= 1) return 1;
  return value;
}
