// @a0n/.ignored_gnosis/distributed-inference-host/src/image-pad.ts
// Mirror-padding eliminates the FFT-circular wrap halo on non-periodic images.
// Hot path for deblur-any-image.ts. Mirror reflection (a, b, c, d -> a, b, c, d, c, b, a, b, c, d, ...)
// is the choice over zero-padding because zero-padding creates a step
// discontinuity at the seam that the deblur amplifies into ringing -
// exactly the artifact we are trying to suppress.

/**
 * Handles the gnosis next Pow2 workflow.
 */
export function nextPow2(n: number): number {
  if (!Number.isFinite(n) || !Number.isInteger(n) || n < 1) {
    throw new RangeError(`nextPow2: n must be a positive integer, got ${n}`);
  }
  let p = 1;
  while (p < n) p <<= 1;
  return p;
}

export interface PadResult {
  plane: Float32Array;       // length paddedWidth * paddedHeight
  paddedWidth: number;       // power of 2
  paddedHeight: number;      // power of 2
  originX: number;           // x offset where original (0,0) lives in padded plane
  originY: number;           // y offset where original (0,0) lives in padded plane
  originalWidth: number;
  originalHeight: number;
}

export interface MirrorPadArgs {
  plane: Float32Array;       // length width * height (single channel)
  width: number;
  height: number;
  paddedWidth?: number;      // defaults to nextPow2(width)
  paddedHeight?: number;     // defaults to nextPow2(height)
}

function isPow2(n: number): boolean {
  return Number.isInteger(n) && n > 0 && (n & (n - 1)) === 0;
}

// Standard symmetric (whole-sample) mirror reflection.
// d is the offset from the original-origin in the padded plane (can be negative).
// n is the original size along this axis.
function reflect(d: number, n: number): number {
  if (n <= 1) return 0;
  const period = 2 * n - 2;
  let m = ((d % period) + period) % period;
  return m < n ? m : period - m;
}

/**
 * Handles the gnosis mirror Pad workflow.
 */
export function mirrorPad(args: MirrorPadArgs): PadResult {
  const { plane, width, height } = args;
  if (!Number.isInteger(width) || width <= 0) {
    throw new RangeError(`mirrorPad: width must be a positive integer, got ${width}`);
  }
  if (!Number.isInteger(height) || height <= 0) {
    throw new RangeError(`mirrorPad: height must be a positive integer, got ${height}`);
  }
  if (plane.length !== width * height) {
    throw new RangeError(
      `mirrorPad: plane.length (${plane.length}) != width*height (${width * height})`,
    );
  }

  const paddedWidth = args.paddedWidth ?? nextPow2(width);
  const paddedHeight = args.paddedHeight ?? nextPow2(height);

  if (!isPow2(paddedWidth)) {
    throw new RangeError(`mirrorPad: paddedWidth must be a power of 2, got ${paddedWidth}`);
  }
  if (!isPow2(paddedHeight)) {
    throw new RangeError(`mirrorPad: paddedHeight must be a power of 2, got ${paddedHeight}`);
  }
  if (paddedWidth < width) {
    throw new RangeError(
      `mirrorPad: paddedWidth (${paddedWidth}) must be >= width (${width})`,
    );
  }
  if (paddedHeight < height) {
    throw new RangeError(
      `mirrorPad: paddedHeight (${paddedHeight}) must be >= height (${height})`,
    );
  }

  const originX = Math.floor((paddedWidth - width) / 2);
  const originY = Math.floor((paddedHeight - height) / 2);

  // Pre-compute source-x and source-y indices once each.
  const srcX = new Int32Array(paddedWidth);
  for (let i = 0; i < paddedWidth; i++) {
    srcX[i] = reflect(i - originX, width);
  }
  const srcY = new Int32Array(paddedHeight);
  for (let j = 0; j < paddedHeight; j++) {
    srcY[j] = reflect(j - originY, height);
  }

  const out = new Float32Array(paddedWidth * paddedHeight);
  for (let j = 0; j < paddedHeight; j++) {
    const sy = srcY[j];
    const rowSrc = sy * width;
    const rowDst = j * paddedWidth;
    for (let i = 0; i < paddedWidth; i++) {
      out[rowDst + i] = plane[rowSrc + srcX[i]];
    }
  }

  return {
    plane: out,
    paddedWidth,
    paddedHeight,
    originX,
    originY,
    originalWidth: width,
    originalHeight: height,
  };
}

export interface CropCenterArgs {
  plane: Float32Array;        // length paddedWidth * paddedHeight
  paddedWidth: number;
  paddedHeight: number;
  originX: number;
  originY: number;
  cropWidth: number;
  cropHeight: number;
}

// Inverse of mirrorPad: extracts a (cropWidth, cropHeight) rectangle starting
// at (originX, originY) from the padded plane.
/**
 * Handles the gnosis crop Center workflow.
 */
export function cropCenter(args: CropCenterArgs): Float32Array {
  const {
    plane,
    paddedWidth,
    paddedHeight,
    originX,
    originY,
    cropWidth,
    cropHeight,
  } = args;

  if (!Number.isInteger(paddedWidth) || paddedWidth <= 0) {
    throw new RangeError(`cropCenter: paddedWidth must be a positive integer, got ${paddedWidth}`);
  }
  if (!Number.isInteger(paddedHeight) || paddedHeight <= 0) {
    throw new RangeError(`cropCenter: paddedHeight must be a positive integer, got ${paddedHeight}`);
  }
  if (plane.length !== paddedWidth * paddedHeight) {
    throw new RangeError(
      `cropCenter: plane.length (${plane.length}) != paddedWidth*paddedHeight (${paddedWidth * paddedHeight})`,
    );
  }
  if (!Number.isInteger(cropWidth) || cropWidth <= 0) {
    throw new RangeError(`cropCenter: cropWidth must be a positive integer, got ${cropWidth}`);
  }
  if (!Number.isInteger(cropHeight) || cropHeight <= 0) {
    throw new RangeError(`cropCenter: cropHeight must be a positive integer, got ${cropHeight}`);
  }
  if (!Number.isInteger(originX) || originX < 0) {
    throw new RangeError(`cropCenter: originX must be a non-negative integer, got ${originX}`);
  }
  if (!Number.isInteger(originY) || originY < 0) {
    throw new RangeError(`cropCenter: originY must be a non-negative integer, got ${originY}`);
  }
  if (originX + cropWidth > paddedWidth) {
    throw new RangeError(
      `cropCenter: originX (${originX}) + cropWidth (${cropWidth}) > paddedWidth (${paddedWidth})`,
    );
  }
  if (originY + cropHeight > paddedHeight) {
    throw new RangeError(
      `cropCenter: originY (${originY}) + cropHeight (${cropHeight}) > paddedHeight (${paddedHeight})`,
    );
  }

  const out = new Float32Array(cropWidth * cropHeight);
  for (let j = 0; j < cropHeight; j++) {
    const rowSrc = (originY + j) * paddedWidth + originX;
    const rowDst = j * cropWidth;
    for (let i = 0; i < cropWidth; i++) {
      out[rowDst + i] = plane[rowSrc + i];
    }
  }
  return out;
}
