// @a0n/.ignored_gnosis/distributed-inference-host/src/image-spectrum-classify.ts
// Host-side metadata classification layered on top of `CarrierMediumReadingKind`.
// The carrier layer stays binary ('evolution' | 'resolution') — we do NOT add a
// third kind there; instead this module attaches a Bule-face label to a single
// deblur step at the host. Precedence is opportunity > waste > diversity because
// Wiener with a known PSF should produce evolution steps; if the smoke ever lands
// on waste or diversity, that signals the regularization has flattened detail
// rather than recovered it. Phase 1 is single-shot (one Wiener step → one
// classification); the API is iteration-shaped so Phase 2 can swap in iterative
// deblur without touching callers.

import { fft2 } from './fft2';
import type { BuleyFace } from './spectral-noise-monitor';

export interface SpectrumBands {
  totalEnergy: number;
  lowBandEnergy: number;
  highBandEnergy: number;
}

/**
 * Handles the gnosis measure Spectrum Bands workflow.
 */
export function measureSpectrumBands(
  plane: Float32Array,
  w: number,
  h: number,
  lowFreqCutoff?: number,
): SpectrumBands {
  const cutoff = lowFreqCutoff ?? Math.min(w, h) / 8;
  const spec = fft2(plane, w, h);
  const re = spec.re;
  const im = spec.im;

  let totalEnergy = 0;
  let lowBandEnergy = 0;
  let highBandEnergy = 0;

  const halfW = w / 2;
  const halfH = h / 2;

  for (let ky = 0; ky < h; ky++) {
    const kySigned = ky <= halfH ? ky : ky - h;
    const absKy = Math.abs(kySigned);
    const rowOffset = ky * w;
    for (let kx = 0; kx < w; kx++) {
      // DC is a sample-mean offset, not a frequency — skip from BOTH bands and total.
      if (kx === 0 && ky === 0) continue;
      const kxSigned = kx <= halfW ? kx : kx - w;
      const absKx = Math.abs(kxSigned);
      const radius = absKx > absKy ? absKx : absKy;

      const idx = rowOffset + kx;
      const r = re[idx];
      const i = im[idx];
      const mag2 = r * r + i * i;

      totalEnergy += mag2;
      if (radius < cutoff) {
        lowBandEnergy += mag2;
      } else {
        highBandEnergy += mag2;
      }
    }
  }

  return { totalEnergy, lowBandEnergy, highBandEnergy };
}

export interface DeblurStepClassification {
  face: BuleyFace;
  lowBandDelta: number;
  highBandDelta: number;
  totalDelta: number;
  reason: 'evolution' | 'resolution' | 'shuffle';
}

export interface ClassifyDeblurStepArgs {
  before: Float32Array;
  after: Float32Array;
  w: number;
  h: number;
  lowFreqCutoff?: number;
  energyThreshold?: number;
}

/**
 * Handles the gnosis classify Deblur Step workflow.
 */
export function classifyDeblurStep(
  args: ClassifyDeblurStepArgs,
): DeblurStepClassification {
  const { before, after, w, h, lowFreqCutoff, energyThreshold } = args;

  const beforeBands = measureSpectrumBands(before, w, h, lowFreqCutoff);
  const afterBands = measureSpectrumBands(after, w, h, lowFreqCutoff);

  const lowBandDelta = afterBands.lowBandEnergy - beforeBands.lowBandEnergy;
  const highBandDelta = afterBands.highBandEnergy - beforeBands.highBandEnergy;
  const totalDelta = afterBands.totalEnergy - beforeBands.totalEnergy;

  // tau is computed AFTER measuring `before`, against before.totalEnergy.
  const tau = (energyThreshold ?? 1e-3) * beforeBands.totalEnergy;

  // Precedence: opportunity > waste > diversity.
  // Negative lowBandDelta is meaningful — do not clamp deltas.
  let face: BuleyFace;
  let reason: 'evolution' | 'resolution' | 'shuffle';
  if (highBandDelta > tau) {
    face = 'opportunity';
    reason = 'evolution';
  } else if (lowBandDelta > 0 && highBandDelta >= -tau) {
    face = 'waste';
    reason = 'resolution';
  } else {
    face = 'diversity';
    reason = 'shuffle';
  }

  return { face, lowBandDelta, highBandDelta, totalDelta, reason };
}
