// @a0n/.ignored_gnosis/distributed-inference-host/src/image-rgb-frf.ts
// File header (5-7 lines, MANDATORY): explicit FRF over the three colour
// channels of an interleaved RGB Float32 plane. FORK runs the caller-
// supplied per-channel transform on each plane in parallel (async-shaped;
// CPU-bound forks serialise on a single Node thread but the structure is
// the canonical pattern). RACE is identity. FOLD interleaves R, G, B back
// into a single Float32Array(width*height*3).

import { forkRaceFold, type ForkTask } from './frf';

export type ChannelTransform = (
  channel: Float32Array,
  width: number,
  height: number,
  channelIndex: 0 | 1 | 2,
) => Promise<Float32Array> | Float32Array;

export interface RGBSplitArgs {
  /** Interleaved RGB Float32, length = width*height*3, row-major. */
  data: Float32Array;
  width: number;
  height: number;
}

export interface RGBJoinArgs {
  channels: ReadonlyArray<Float32Array>; // [R, G, B], each length width*height
  width: number;
  height: number;
}

export interface RGBFRFArgs {
  data: Float32Array;
  width: number;
  height: number;
  /** Per-channel transform — receives one Float32Array, returns one. */
  transform: ChannelTransform;
  /** Concurrency cap for the forkRaceFold. Default 3. */
  concurrency?: number;
}

function validateDims(width: number, height: number, fn: string): void {
  if (!Number.isFinite(width) || !Number.isInteger(width) || width <= 0) {
    throw new RangeError(`${fn}: width must be a positive integer (got ${width})`);
  }
  if (!Number.isFinite(height) || !Number.isInteger(height) || height <= 0) {
    throw new RangeError(`${fn}: height must be a positive integer (got ${height})`);
  }
}

/** Split an interleaved RGB plane into 3 single-channel Float32Arrays. */
export function splitRGB(
  args: RGBSplitArgs,
): [Float32Array, Float32Array, Float32Array] {
  const { data, width, height } = args;
  validateDims(width, height, 'splitRGB');
  const expected = width * height * 3;
  if (data.length !== expected) {
    throw new RangeError(
      `splitRGB: data.length (${data.length}) !== width*height*3 (${expected})`,
    );
  }
  const n = width * height;
  const r = new Float32Array(n);
  const g = new Float32Array(n);
  const b = new Float32Array(n);
  for (let i = 0; i < n; i++) {
    const j = i * 3;
    r[i] = data[j];
    g[i] = data[j + 1];
    b[i] = data[j + 2];
  }
  return [r, g, b];
}

/** Join 3 channel planes back into an interleaved RGB Float32. */
export function joinRGB(args: RGBJoinArgs): Float32Array {
  const { channels, width, height } = args;
  validateDims(width, height, 'joinRGB');
  if (channels.length !== 3) {
    throw new RangeError(
      `joinRGB: channels must have length 3 (got ${channels.length})`,
    );
  }
  const n = width * height;
  for (let c = 0; c < 3; c++) {
    if (channels[c].length !== n) {
      throw new RangeError(
        `joinRGB: channels[${c}].length (${channels[c].length}) !== width*height (${n})`,
      );
    }
  }
  const [r, g, b] = channels;
  const out = new Float32Array(n * 3);
  for (let i = 0; i < n; i++) {
    const j = i * 3;
    out[j] = r[i];
    out[j + 1] = g[i];
    out[j + 2] = b[i];
  }
  return out;
}

/** FORK 3 channels through `transform` in parallel; RACE identity;
 *  FOLD interleave back to RGB. Returns the recovered interleaved plane. */
export async function rgbFRF(args: RGBFRFArgs): Promise<Float32Array> {
  const { data, width, height, transform } = args;
  const concurrency = args.concurrency ?? 3;

  validateDims(width, height, 'rgbFRF');
  const expected = width * height * 3;
  if (data.length !== expected) {
    throw new RangeError(
      `rgbFRF: data.length (${data.length}) !== width*height*3 (${expected})`,
    );
  }
  if (typeof transform !== 'function') {
    throw new RangeError(
      `rgbFRF: transform must be a function (got ${typeof transform})`,
    );
  }
  if (
    !Number.isFinite(concurrency) ||
    !Number.isInteger(concurrency) ||
    concurrency < 1
  ) {
    throw new RangeError(
      `rgbFRF: concurrency must be an integer >= 1 (got ${concurrency})`,
    );
  }

  const [rPlane, gPlane, bPlane] = splitRGB({ data, width, height });
  const planes: ReadonlyArray<Float32Array> = [rPlane, gPlane, bPlane];

  const forks: ReadonlyArray<ForkTask<Float32Array>> = [0, 1, 2].map(
    (idx) => async () => {
      const ch = planes[idx];
      const result = await Promise.resolve(
        transform(ch, width, height, idx as 0 | 1 | 2),
      );
      if (!(result instanceof Float32Array)) {
        throw new TypeError(
          `rgbFRF: transform(channel=${idx}) must return Float32Array`,
        );
      }
      if (result.length !== ch.length) {
        throw new RangeError(
          `rgbFRF: transform(channel=${idx}) returned length ${result.length}, expected ${ch.length}`,
        );
      }
      return result;
    },
  );

  const { output } = await forkRaceFold<Float32Array, Float32Array>({
    forks,
    fold: (survivors) =>
      joinRGB({
        channels: [survivors[0], survivors[1], survivors[2]],
        width,
        height,
      }),
    concurrency,
  });

  return output;
}
