// @a0n/.ignored_gnosis/distributed-inference-host/src/index.ts
/**
 * @a0n/distributed-inference-host
 *
 * Workspace-shared shim wrapping the wasm32 build of
 * open-source/gnosis/distributed-inference. Provides:
 *   - knot file header parsing
 *   - R2 / HTTP / FS Range fetching via a single KnotByteSource interface
 *   - layer prefetch with LRU eviction
 *   - the orchestration sequence between async fetches and sync WASM calls
 *
 * Every Worker that runs distributed inference imports this package so the
 * orchestration logic stays in one place.
 */

export { HostShim } from './host-shim';
export { extractMelSpectrogram, WHISPER_MEL } from './mel-spectrogram';
export {
  STATELESS_EXTRUSION_WGSL,
  STATELESS_MESH_PROJECTION_SCHEMA_VERSION,
  buildAlphaBoundaryWallIndices,
  buildStatelessMeshGrid,
  validateProjectionTextureContract,
} from './stateless-mesh-projection';
export type {
  AlphaBoundaryOptions,
  ProjectionTextureContract,
  ProjectionTextureValidation,
  StatelessMeshGridBuffers,
  StatelessMeshGridOptions,
} from './stateless-mesh-projection';
export {
  parseKnotHeaderFromBytes,
  readKnotHeader,
  computePayloadOffset,
  type KnotHeader,
} from './knot-header';
export { LayerLRU } from './lru';
export {
  PER_LAYER_TENSOR_SUFFIXES,
  RESIDENT_TENSOR_NAMES,
  tensorNamesForLayer,
} from './tensor-names';
export type {
  HostShimOptions,
  KnotByteSource,
  KnotMetadata,
  TensorInfo,
  WasmHostBackendLike,
  WasmLlamaPipelineLike,
  WasmModule,
} from './types';

// Wallington Rotation surfaces (chunked log-rolling pipelined prefill).
export { Station } from './station';
export type { StationConfig, StationEndpoint } from './station';
// UDS variant — local-only, transport-cut version of Station. Backed by
// `fat-station-uds` (open-source/gnosis/distributed-inference). See
// docs/pneuma/local-mesh-uds.md for the wire format.
export { StationUds } from './station-uds';
export type { StationUdsConfig } from './station-uds';
export { StationFlowUdp } from './station-flow-udp';
// Transport-agnostic surface: every Station* class implements `IStation`.
// New code should hold `IStation` (not the concrete class) and prefer
// `openStation()` over the per-class constructors. See
// docs/pneuma/transport-selector.md for the migration roadmap.
export type { AckermannAdmissionGate, IStation } from './types';
export { hasHealthProbe, type IStationWithHealth } from './station-like';
export {
  openStation,
  type OpenStationSpec,
  type StationRole,
  type TransportFactories,
} from './transport-selector';
export { Pipeline } from './pipeline';
export type { PipelineConfig, PrefillResult } from './pipeline';
export {
  AXIS_SHARD_CAPACITY,
  MAHA_ANCHOR,
  PHYLE_AXIS_COUNT,
  getShardAxis,
  mapToMahaShard,
  resolveMahaPhysicalStationRange,
  resolveMahaShardRoute,
  resolveMahaShardedStations,
} from './amplituhedron-sharding';
export type {
  MahaPhysicalStationRange,
  MahaShardRoute,
} from './amplituhedron-sharding';
export {
  PrefillWindowManager,
  prefillTelemetryHeaders,
} from './prefill-window';
export type {
  PrefillGenerationResult,
  PrefillMessage,
  PrefillTelemetry,
  PrefillTelemetryStatus,
  PrefillWindowSnapshot,
  PrefillWindowState,
} from './prefill-window';
export {
  SuperserverMcpClient,
  createSuperserverMcpClient,
  formatToolsForPrompt,
  renderMessagesForPrompt,
  runAgenticChatCompletion,
} from './agentic-chat';
export type {
  AgenticChatCompletionResponse,
  AgenticChatMessage,
  AgenticChatRequest,
  AgenticChatRuntime,
  AgenticMcpClient,
  AgenticToolCall,
  AgenticToolChoice,
  AgenticToolDefinition,
  AgenticTraceEntry,
  McpPreflightResult,
  SuperserverMcpClientConfig,
  TextGenerationRequest,
  TextGenerationResult,
  ToolActivationMetadata,
  ToolExecutionResult,
} from './agentic-chat';
export {
  UNLIMITED_OCR_ARTIFACTS,
  UNLIMITED_OCR_CHAT_TEMPLATE,
  UNLIMITED_OCR_DEFAULT_PROMPT,
  UNLIMITED_OCR_DEFAULT_QUANT,
  UNLIMITED_OCR_MMPROJ_FILENAME,
  UNLIMITED_OCR_MODEL_FILENAME,
  UNLIMITED_OCR_MODEL_ID,
  UNLIMITED_OCR_REPO_ID,
  buildUnlimitedOcrChatRequest,
  validateUnlimitedOcrOutput,
} from './unlimited-ocr-qspec';
export type {
  UnlimitedOcrArtifactSpec,
  UnlimitedOcrChatCompletionRequest,
  UnlimitedOcrChatContentPart,
  UnlimitedOcrProbeSpec,
} from './unlimited-ocr-qspec';
export {
  inferDocumentKind,
  runUnlimitedOcrDocument,
} from './unlimited-ocr-runner';
export type {
  UnlimitedOcrCommandOutput,
  UnlimitedOcrCommandRun,
  UnlimitedOcrCommandRunner,
  UnlimitedOcrDocumentInput,
  UnlimitedOcrDocumentKind,
  UnlimitedOcrDocumentResult,
  UnlimitedOcrRunnerOptions,
} from './unlimited-ocr-runner';
export {
  AEON_STREAM_PNEUMA_INTERNAL,
  THOTH_GATE_KINDS,
  THOTH_GATE_TELEMETRY_FRAME_KIND,
  canonicalThothGateTelemetryMetrics,
  encodeThothGateTelemetryFrame,
  isThothGateTelemetryFrame,
  renderThothGateTelemetryPromptContext,
  thothGateTelemetryActionHint,
  thothGateTelemetryReadyHandoffCount,
  thothGateTelemetryTotalPressure,
} from './thoth-gate-telemetry';
export type {
  ThothGateKind,
  ThothGateScore,
  ThothGateTelemetryFrameWire,
  ThothGateTelemetryMetrics,
} from './thoth-gate-telemetry';
export {
  THOTH_MIND_BODY_SPIRIT_PROMPT_FEEDBACK_REJECTION_RESIDUE_THEOREM,
  THOTH_MIND_BODY_SPIRIT_SCRIBE_ADMISSIBILITY_THEOREM,
  THOTH_PROMPT_FEEDBACK_THEOREM,
  assertThothPromptFeedbackResponse,
  isThothPromptFeedbackResponse,
  validateThothPromptFeedbackResponse,
} from './thoth-prompt-feedback';
export type {
  ThothPromptFeedbackAdmissionRejection,
  ThothPromptFeedbackRequest,
  ThothPromptFeedbackResponse,
  ThothPromptFeedbackValidation,
} from './thoth-prompt-feedback';
export {
  THOTH_OATH_AUDIT_TOOL,
  THOTH_OATH_CLEAR_TOOL,
  THOTH_OATH_PROPOSE_TOOL,
  THOTH_OATH_TOOL_DEFINITIONS,
  ThothOathSelfManagementMcpClient,
  createThothOathSelfManagementMcpClient,
} from './thoth-oath-tools';
export {
  ThothOathFileLedgerStore,
  createThothOathFileLedgerStore,
} from './thoth-oath-ledger-file-store';
export { ThothOathClient } from './thoth-oath-client';
export type {
  ThothOathClientConfig,
  ThothOathCreateInput,
  ThothOathCreateResponse,
  ThothOathDeleteResponse,
  ThothOathFetcher,
  ThothOathLedgerResponse,
  ThothOathRuntimeRecord,
  ThothOathUpdateInput,
  ThothOathUpdateResponse,
} from './thoth-oath-client';
export type {
  ThothOathLedgerStore,
  ThothOathRecord,
  ThothOathSchedule,
  ThothOathScheduleMode,
  ThothOathTelemetryEvent,
  ThothOathTelemetryEventKind,
  ThothOathToolOptions,
} from './thoth-oath-tools';
export {
  AEON_EIGHT_HEAD_OPTIMIZER_REQUIRES_LIFT_THEOREM,
  AEON_EIGHT_HEAD_OPTIMIZER_THEOREM_LINEAGE,
  ALIASED_OPTIMIZER_THEOREM_LINEAGE,
  LIFTED_EIGHT_HEAD_OPTIMIZER_ADMITS_RUNTIME_WORK_THEOREM,
  LIFTED_EIGHT_HEAD_OPTIMIZER_THEOREM_LINEAGE,
  LIFTED_PINK_EIGHT_HEAD_OPTIMIZER_ADMITS_RUNTIME_WORK_THEOREM,
  LIFTED_PINK_EIGHT_HEAD_OPTIMIZER_THEOREM_LINEAGE,
  OPTIMIZER_ALIAS_REQUIRES_RESOLUTION_LIFT_THEOREM,
  OPTIMIZER_READY_ADMITS_RUNTIME_WORK_THEOREM,
  OPTIMIZER_RESIDUE_REQUIRES_RESOLUTION_LIFT_THEOREM,
  READY_OPTIMIZER_THEOREM_LINEAGE,
  RESIDUE_OPTIMIZER_THEOREM_LINEAGE,
  SpectralNoiseMonitor,
  TOWER_CEILINGS,
  UNPROVED_THRESHOLD_POLICY_THEOREM_LINEAGE,
  actionFace,
  classifyMeshAttentionClosure,
  classifySpectralNoise,
  clinamenContract,
  clinamenLift,
  cyclePermute,
  entropyFace,
  estimateMeshAttentionClosureSavedWork,
  fingerprintBuleyProjection,
  governanceSelfSimilarityViolation,
  measureBuley,
  meshAttentionClosureTheoremLineage,
  meshAttentionClosureWorkPolicy,
  reduceTraceFingerprint,
  requestEntropy,
  selfSimilarityViolation,
  vacuumBuleyProjection,
  wasteFace,
} from './spectral-noise-monitor';
export type {
  LayerNodeTrace,
  AttentionClosureInput,
  AttentionClosureKind,
  AttentionClosureState,
  BuleyFace,
  BuleyProjection,
  DarkDeceptaconResolutionLiftDecision,
  MeshAttentionClosureAssessment,
  MeshAttentionClosureConfig,
  MeshAttentionClosureTheorem,
  MeshAttentionClosureWorkEstimate,
  MeshAttentionClosureWorkload,
  MeshAttentionClosureWorkPolicy,
  ResolutionLiftActionForMesh,
  SelfSimilarityViolation,
  SpectralNoiseAction,
  SpectralNoiseAssessment,
  SpectralNoiseColor,
  SpectralNoiseFingerprint,
  SpectralNoiseMonitorConfig,
  SpectralNoiseRuntimeState,
} from './spectral-noise-monitor';
export {
  HORN_OBSERVABILITY_RUNTIME_SCHEMA_VERSION,
  HORN_OBSERVABILITY_THEOREM_LINEAGE,
  computeHornObservabilityFromSpectralAssessment,
  computeHornObservabilityRuntime,
} from './horn-observability-runtime';
export type {
  HornObservabilityRuntimeInput,
  HornObservabilityRuntimeSource,
  HornObservabilityRuntimeWire,
} from './horn-observability-runtime';
export {
  MESH_CARRIER_ASSESSMENT_CERTIFICATE_VERSION,
  MESH_CARRIER_EPISTEMIC_LIMITATION_CLASSES,
  MESH_CARRIER_EPISTEMIC_OUTCOMES,
  MESH_HTTP_HEADER_CARRIER_DEBLUR_FOLD,
  MESH_HTTP_HEADER_CARRIER_STYLE_FOLD,
  attachDeblurEnsembleToMeshCarrierAssessmentJson,
  attachStyleTransferEnsembleToMeshCarrierAssessmentJson,
  attentionClosureAssessmentToWire,
  computeMeshCarrierNoiseSignalFromAssessments,
  deblurEnsembleResultToMeshCarrierFold,
  deriveMeshCarrierEpistemicFromSpectral,
  encodeDeblurFoldWireForMeshCarrierRequestHeader,
  encodeMeshCarrierAssessmentV1Json,
  encodeStyleTransferFoldWireForMeshCarrierRequestHeader,
  mergeDeblurFoldIntoMeshCarrierJson,
  mergeStyleTransferFoldIntoMeshCarrierJson,
  meshCarrierRequestHeadersWithDeblurEnsemble,
  meshCarrierRequestHeadersWithStyleTransferEnsemble,
  styleTransferEnsembleResultToMeshCarrierFold,
  toMeshCarrierAssessmentWire,
} from './mesh-carrier-assessment';
export type {
  EncodeMeshCarrierAssessmentV1Args,
  MeshCarrierAssessmentV1Wire,
  MeshCarrierDeblurFoldWire,
  MeshCarrierEpistemicLimitationClass,
  MeshCarrierEpistemicOutcome,
  MeshCarrierEpistemicWire,
  MeshCarrierNoiseSignalInterpretation,
  MeshCarrierNoiseSignalProjection,
  MeshCarrierStyleTransferFoldWire,
} from './mesh-carrier-assessment';
export {
  applyInvisibleWatermarkToPlane,
  resolveInvisibleWatermarkOptions,
  watermarkResidualCorrelation,
} from './invisible-watermark';
export type {
  InvisibleWatermarkMode,
  InvisibleWatermarkOptions,
  InvisibleWatermarkProvenance,
  InvisibleWatermarkResolved,
} from './invisible-watermark';
export {
  admittedVisualWatermarkToken,
  admitVisualWatermarkMeshIdentity,
  visualWatermarkMeshIdentityMasterRuntime,
} from './visual-watermark-mesh-identity';
export type {
  RecoveredVisualWatermark,
  VisualWatermarkMeshEpistemicOutcome,
  VisualWatermarkMeshIdentityAdmission,
  VisualWatermarkMeshIdentityAdmitOptions,
  VisualWatermarkMeshLimitationClass,
} from './visual-watermark-mesh-identity';
export {
  styleTransferAnyImage,
  encodeStyledAsPNG,
  runClassicalStyleTransfer,
  runAdaINAdapter,
  forkRaceFoldStyleTransfer,
  compositeStyleTransferFoldScore,
  resolveTopologyAlphaForStyle,
  stubTopologyAlphaFromContentRgb,
  topologyAlphaStyleBackendFromImagingWasm,
  tryWasmMeshDepthConfidence,
  STYLE_TRANSFER_MESH_GATE_SCHEMA_VERSION,
  STYLE_TRANSFER_MESH_SIX_CHANNEL_IDS,
  STYLE_TRANSFER_TOPOLOGY_GLUE_SCHEMA_VERSION,
} from './style-public';
export type {
  AdaINAdapterBackend,
  AdaINAdapterRequest,
  TopologyAlphaStyleBackend,
  TopologyAlphaStyleRequest,
  TopologyAlphaStyleResult,
  StyleTransferImagingWasmSurface,
  StyleTransferAnyImageOptions,
  StyleTransferAnyImageResult,
  StyleTransferCandidateReport,
  StyleTransferChannelStats,
  StyleTransferEnsembleArgs,
  StyleTransferEnsembleConfig,
  StyleTransferEnsembleResult,
  StyleTransferImageInput,
  StyleTransferMeshAdmissionChannel,
  StyleTransferMeshFoldMetrics,
  StyleTransferMeshProvenance,
  StyleTransferMethod,
} from './style-public';
export {
  dynamicBranch,
  manifoldSaturation,
  mapInstruction,
  mapPolyglot,
  pathComplexity,
  resonanceAlpha,
} from './universal-signal-map';
export type {
  MonsterPath,
  PolyglotOp,
  ResonanceColor,
  SignalToken,
} from './universal-signal-map';
export {
  buleyCostAlgebra,
  buleyFaceUnit,
  composeHom,
  composeWithFaceUnit,
  idHom,
  natCostAlgebra,
  scoreHom,
  validateCostHom,
} from './cost-algebra';
export type { CostAlgebra, CostHom } from './cost-algebra';
export {
  buleHotReplicaAdmission,
  clonedPairScore,
  correctiveContractsForReplica,
  diagonal,
  entropyGeneratedByCloning,
  hotReplicaAdmission,
  isDiagonalScorePreserving,
  replicaTotalHeat,
  replicationEntropy,
} from './cost-algebra-entropy';
export type { HotReplicaAdmission } from './cost-algebra-entropy';
export {
  bigBangReach,
  pillarFutureAnchor,
  pillarPastAnchor,
  pillarPresentAnchor,
  threePillarReport,
} from './pillar-anchors';
export type {
  PillarFutureResult,
  PillarPresentResult,
  ThreePillarReport,
} from './pillar-anchors';
export {
  CostAlgebraViolation,
  requireBuleHotReplicaAdmission,
  requireCostHom,
  requireHotReplicaAdmission,
  requireOperationalPillars,
  requirePillarFuture,
  requirePillarPast,
  requirePillarPresent,
  requireThreePillars,
} from './cost-algebra-enforcement';
export {
  costlyBroadcast,
  smartBroadcast,
  vacuumBroadcast,
} from './cost-algebra-broadcast';
export type {
  BroadcastResult,
  CostlyBroadcastRequest,
  CostlyBroadcastResult,
  VacuumBroadcastRequest,
  VacuumBroadcastResult,
} from './cost-algebra-broadcast';
export {
  CostLedger,
  FRF,
  GROUNDING_TRITON,
  MonsterCache,
  PLEROMATIC_CLOSURE,
  SHARED_UNIT,
  SovereignJudge,
  TRIPLE_POINT,
  TRITON_STRETCH_FACTOR,
  groundingTritonSnapshot,
} from './pleromatic-core';
export type { DescentVerification, GroundingNode } from './pleromatic-core';
export {
  LENS_FACTOR_HEXON_TO_TRIHEXENNEON,
  LensingMaster,
  TRIHEXENNEON_IS_NINE_HEXONS,
  lensAddress,
  lensAddressReconstructRoundtrip,
  lensDescent,
  lensPreservesUnitCost,
  lensReconstruct,
  lensReconstructAddressRoundtrip,
  tritonStretchDivRecovers,
  tritonStretchInjective,
} from './pleromatic-lensing';
export type { LensAddress } from './pleromatic-lensing';
export {
  BANDWIDTH_HEXON,
  BANDWIDTH_TRIHEXENNEON,
  BANDWIDTH_TRITON,
  EvolutionResolutionMaster,
  OneWayMirrorMaster,
  bandwidthGapStrict,
  buleStepCost,
  buleStepCostInvariantUnderBandwidthChange,
  carrierBandwidth,
  copyIndexExceedsTritonBandwidth,
  eighteenToOneCollision,
  evolutionTriggerAtBandwidthMinusOne,
  isEvolutionStep,
  isResolutionStep,
  lensDownToTriton,
  lensTritonThreeFibers,
  lowerObserverCannotDistinguishFirstTriton,
  stepEitherOr,
  stepNotBoth,
  tritonStretchTriplesBandwidth,
} from './pleromatic-bandwidth';
export {
  CarrierMediumPipeline,
  analyzeCarrierMedium,
  carrierFiberForPosition,
  classifyCarrierStep,
  projectCarrierPosition,
  quantizeUnitToCarrierStep,
} from './carrier-medium';
export type {
  CarrierFiber,
  CarrierMediumAnalysis,
  CarrierMediumBandwidthSummary,
  CarrierMediumCell,
  CarrierMediumCellInput,
  CarrierMediumInput,
  CarrierMediumPipelineConfig,
  CarrierMediumReading,
  CarrierMediumReadingKind,
} from './carrier-medium';
export { ImageCarrierPipeline } from './image-carrier-pipeline';
export type {
  ImageCarrierAnalysis,
  ImageCarrierBandwidthSummary,
  ImageCarrierInput,
  ImageCarrierPatch,
  ImageCarrierPipelineConfig,
  ImageCarrierReading,
  ImageCarrierReadingKind,
  ImageSampleRange,
} from './image-carrier-pipeline';
export {
  LIGHT_FIELD_HEIGHT,
  LIGHT_FIELD_WIDTH,
  LIGHT_MAX_WAVE_MODES,
  LIGHT_MODEL_BUNDLE_SCHEMA,
  LIGHT_MODEL_BUNDLE_VERSION,
  LIGHT_MODEL_DESCRIPTION,
  clampLight01,
  createLightModelBundle,
  lightBase64ToBytes,
  lightBytesToBase64,
  parseLightModelBundle,
  parseLightMonsterMeshArtifact,
  projectLightMonsterMeshArtifact,
  reconstructLightModelBundle,
} from './light-model-artifact';
export type {
  LightFoveaPoint,
  LightModelBundle,
  LightMonsterMeshArtifact,
  LightReverseReconstruction,
  LightStandingWaveMode,
  LightTranslationArtifactInput,
  LightTranslationMetrics,
  LightTranslationSettings,
} from './light-model-artifact';
export {
  MCKAY_C1,
  MCKAY_C2,
  MCKAY_CHI1,
  MCKAY_CHI2,
  MCKAY_CHI3,
  PleromaticMonsterMeshMaster,
  TRITON_ROTATION_ORDER,
  clinamenAppearsInMckayC1,
  clinamenAtC1,
  clinamenAtC2,
  forkChildrenConstructiveInterference,
  interferencePatternIsPureCoincidence,
  mckayC1Decomposition,
  mckayC2Decomposition,
  mckayChi1EqualsClinamen,
  tritonRotation,
  tritonRotationOrderThree,
  tritonRotationVisitsAllThree,
} from './pleromatic-monster-mesh';
export {
  ReasoningChainValidator,
  coherenceRatio,
} from './pleromatic-reasoning-validator';
export type {
  ChainVerdict,
  ReasoningStep,
} from './pleromatic-reasoning-validator';
export {
  allStandardModelBosons,
  bosonPhaseCount,
  bosonSkyrmsEquilibrium,
  classifyBosonEquilibrium,
  isHiggsEquilibrium,
} from './boson-skyrms';
export type { StandardModelBoson } from './boson-skyrms';
export {
  COST_ALGEBRA_AXES,
  COUPLING_CONSTANT,
  SUPERSTRING_DIMENSION,
  M_THEORY_DIMENSION,
  BOSONIC_STRING_DIMENSION,
  allStandardModelFermions,
  axisSetTotal,
  baryon,
  darkDoesNotIntersectSM,
  darkPhaseCount,
  darkSkyrmsEquilibrium,
  eleventhDimensionIsCoupling,
  fermionPaysExclusionTax,
  fermionPauliExclusionEntropy,
  fermionPhaseCount,
  fermionSkyrmsEquilibrium,
  isNahmMinimal,
  meson,
  nahmSandwichDimension,
  phaseTransitionDistance,
} from './particle-physics';
export type {
  Antiparticle,
  CompositeParticle,
  DarkSectorParticle,
  FermionFlavor,
  Generation,
  NahmMinimalAxisSet,
  StandardModelFermion,
} from './particle-physics';
export { PneumaPipeline } from './pipeline_pneuma';
export type {
  PneumaPipelineConfig,
  TeleportationConfig,
  TeleportationBootOutcome,
} from './pipeline_pneuma';

// ER=EPR teleportation cache HTTP knob client. Wired against both the wasm
// CF Worker /memo/* surface and fat-station-memo's identical native one.
// See `docs/pneuma/memo-routes.md` for the route table.
export { MemoClient } from './memo-client';
export type {
  MemoClientConfig,
  MemoConfigure,
  MemoStats,
  MemoFetcher,
} from './memo-client';

// Registry failover and cluster facades
export { RegistryShim } from './registry-shim';
export type { RegistryShimConfig } from './registry-shim';
export { DistributedInferenceCluster } from './cluster';
export type { ClusterConfig } from './cluster';
export {
  resolveOpenAiServerMeshConfig,
  buildOpenAiServerMeshPipeline,
  parseOpenAiServerLayerRanges,
} from './openai-server-mesh';
