// @a0n/.ignored_gnosis/distributed-inference-host/src/invisible-watermark.ts
export type InvisibleWatermarkMode = 'luma_spread';

export interface InvisibleWatermarkOptions {
  enabled?: boolean;
  seed?: string;
  strength?: number;
  payload?: string;
  mode?: InvisibleWatermarkMode;
}

export interface InvisibleWatermarkResolved {
  enabled: boolean;
  seed: string;
  strength: number;
  payload: string;
  mode: InvisibleWatermarkMode;
}

export interface InvisibleWatermarkProvenance {
  id: string;
  mode: InvisibleWatermarkMode;
  strength: number;
  seedDigest: string;
  payloadBits: number;
}

const DEFAULT_WATERMARK_MODE: InvisibleWatermarkMode = 'luma_spread';
const DEFAULT_WATERMARK_SEED = 'forkjoin-invisible-watermark-v1';
const DEFAULT_WATERMARK_PAYLOAD = 'deblur-v1';
const DEFAULT_WATERMARK_STRENGTH = 0.02;
const MIN_WATERMARK_STRENGTH = 0.001;
const MAX_WATERMARK_STRENGTH = 0.08;
const WATERMARK_BITS = 64;

/**
 * Resolves the Invisible Watermark Options.
 */
export function resolveInvisibleWatermarkOptions(
  options: InvisibleWatermarkOptions | undefined,
): InvisibleWatermarkResolved {
  const enabled = options?.enabled ?? false;
  const seed = (options?.seed ?? DEFAULT_WATERMARK_SEED).trim() || DEFAULT_WATERMARK_SEED;
  const payload = (options?.payload ?? DEFAULT_WATERMARK_PAYLOAD).trim() || DEFAULT_WATERMARK_PAYLOAD;
  const mode = options?.mode ?? DEFAULT_WATERMARK_MODE;
  const strengthRaw = options?.strength ?? DEFAULT_WATERMARK_STRENGTH;
  if (!Number.isFinite(strengthRaw)) {
    throw new RangeError(`resolveInvisibleWatermarkOptions: strength must be finite, got ${strengthRaw}`);
  }
  const strength = clamp(strengthRaw, MIN_WATERMARK_STRENGTH, MAX_WATERMARK_STRENGTH);
  return { enabled, seed, strength, payload, mode };
}

/**
 * Handles the gnosis apply Invisible Watermark To Plane workflow.
 */
export function applyInvisibleWatermarkToPlane(args: {
  plane: Float32Array;
  width: number;
  height: number;
  channels: 1 | 3;
  options?: InvisibleWatermarkOptions;
  frameIndex?: number;
}): { plane: Float32Array; provenance?: InvisibleWatermarkProvenance } {
  const resolved = resolveInvisibleWatermarkOptions(args.options);
  if (!resolved.enabled) {
    return { plane: args.plane };
  }
  const frameIndex = args.frameIndex ?? 0;
  const taggedPayload = `${resolved.payload}|frame:${frameIndex}`;
  const payloadSeed = `${resolved.seed}|${taggedPayload}|${resolved.mode}`;
  const bits = payloadBits(payloadSeed, WATERMARK_BITS);
  const seedDigest = fnv1a32Hex(resolved.seed);
  const id = fnv1a32Hex(`${seedDigest}|${taggedPayload}|${resolved.mode}|${resolved.strength}`);
  const rng = xorshift32(seedFromString(seedDigest));
  const out = new Float32Array(args.plane);
  const blockSide = 4;
  const blocksX = Math.max(1, Math.floor(args.width / blockSide));
  const blocksY = Math.max(1, Math.floor(args.height / blockSide));
  const maxAnchors = Math.max(1, Math.min(bits.length, blocksX * blocksY));

  for (let bitIdx = 0; bitIdx < maxAnchors; bitIdx++) {
    const bit = bits[bitIdx]!;
    const blockX = Math.floor(rng() * blocksX);
    const blockY = Math.floor(rng() * blocksY);
    const x0 = blockX * blockSide;
    const y0 = blockY * blockSide;
    const sign = bit === 1 ? 1 : -1;
    const amp = sign * resolved.strength;
    for (let yy = 0; yy < blockSide; yy++) {
      const y = y0 + yy;
      if (y >= args.height) continue;
      for (let xx = 0; xx < blockSide; xx++) {
        const x = x0 + xx;
        if (x >= args.width) continue;
        const pix = y * args.width + x;
        if (args.channels === 1) {
          out[pix] = out[pix]! + amp;
        } else {
          const base = pix * 3;
          out[base] = out[base]! + amp;
          out[base + 1] = out[base + 1]! + amp;
          out[base + 2] = out[base + 2]! + amp;
        }
      }
    }
  }

  return {
    plane: out,
    provenance: {
      id,
      mode: resolved.mode,
      strength: resolved.strength,
      seedDigest,
      payloadBits: bits.length,
    },
  };
}

/**
 * Handles the gnosis float Plane To Uint8 Clamped workflow.
 */
export function floatPlaneToUint8Clamped(plane: Float32Array): Uint8Array {
  const out = new Uint8Array(plane.length);
  for (let i = 0; i < plane.length; i++) {
    const v = plane[i]!;
    const c = v <= 0 ? 0 : v >= 1 ? 1 : v;
    out[i] = Math.round(c * 255);
  }
  return out;
}

/**
 * Handles the gnosis watermark Residual Correlation workflow.
 */
export function watermarkResidualCorrelation(args: {
  baseline: Float32Array;
  marked: Float32Array;
  width: number;
  height: number;
  channels: 1 | 3;
  options: InvisibleWatermarkOptions;
  frameIndex?: number;
}): number {
  const resolved = resolveInvisibleWatermarkOptions(args.options);
  if (!resolved.enabled) {
    return 0;
  }
  const synthetic = applyInvisibleWatermarkToPlane({
    plane: args.baseline,
    width: args.width,
    height: args.height,
    channels: args.channels,
    options: resolved,
    frameIndex: args.frameIndex,
  }).plane;
  let dot = 0;
  let normA = 0;
  let normB = 0;
  for (let i = 0; i < args.baseline.length; i++) {
    const expected = synthetic[i]! - args.baseline[i]!;
    const observed = args.marked[i]! - args.baseline[i]!;
    dot += expected * observed;
    normA += expected * expected;
    normB += observed * observed;
  }
  const denom = Math.sqrt(normA * normB);
  return denom > 0 ? dot / denom : 0;
}

function payloadBits(seed: string, bitCount: number): number[] {
  const bits: number[] = [];
  let cursor = seed;
  while (bits.length < bitCount) {
    const hex = fnv1a32Hex(cursor);
    for (let i = 0; i < hex.length && bits.length < bitCount; i++) {
      const nibble = Number.parseInt(hex[i]!, 16);
      for (let b = 0; b < 4 && bits.length < bitCount; b++) {
        bits.push((nibble >> b) & 1);
      }
    }
    cursor = `${cursor}:${hex}`;
  }
  return bits;
}

function fnv1a32Hex(input: string): string {
  let hash = 0x811c9dc5;
  for (let i = 0; i < input.length; i++) {
    hash ^= input.charCodeAt(i);
    hash = Math.imul(hash, 0x01000193);
  }
  return (hash >>> 0).toString(16).padStart(8, '0');
}

function seedFromString(input: string): number {
  return Number.parseInt(fnv1a32Hex(input), 16) >>> 0;
}

function xorshift32(seed: number): () => number {
  let state = seed >>> 0 || 0x9e3779b9;
  return () => {
    state ^= state << 13;
    state ^= state >>> 17;
    state ^= state << 5;
    return (state >>> 0) / 0x1_0000_0000;
  };
}

function clamp(n: number, lo: number, hi: number): number {
  if (n < lo) return lo;
  if (n > hi) return hi;
  return n;
}
