// @a0n/.ignored_gnosis/distributed-inference-host/src/knot-header.ts
/**
 * Knot file header parser. Reads both `.knot` (v1, JSON TOC) and `.kn0t`
 * (v3, binary TOC) — dispatches on the first 4 magic bytes and adapts
 * either into the same `KnotHeader` shape, so every downstream consumer
 * (`KnotModelStorage`, `HostShim`, and transitively the WASM
 * `WasmHostBackend`/`WasmLlamaPipeline`) works unmodified against either
 * format. Full `.kn0t` spec: `open-source/bitwise/KN0T_V3_FORMAT.md`.
 *
 * `.knot` layout (all multi-byte integers little-endian):
 *
 *   bytes 0..4    "KNOT" magic
 *   bytes 4..6    version (u16)
 *   bytes 6..10   metadata length (u32)
 *   bytes 10..    metadata JSON, padded with NUL to a 4-byte boundary
 *   then          1 codec marker byte
 *   then          tensor payload at
 *                 ((10 + meta_len + 3) & ~3) + 1
 *
 * The final `+ 1` accounts for the codec marker byte that sits between the
 * padded metadata block and the tensor payload.
 *
 * `.kn0t` support is a compatibility adapter, not a rewrite of this
 * consumer: it materializes the same flat `{name: {offset, length, type}}`
 * tensors map that `.knot`'s JSON TOC already produces (cheap — decoding an
 * already-parsed binary TOC into ~hundreds of small objects, no
 * `JSON.parse` of a large blob), so this is the ONE place in the serving
 * stack that needs to know `.kn0t` exists. The true zero-copy hot path
 * (`findTensorKn0t`, `readKn0tTocRecord`) stays available in
 * `@a0n/bitwise/kn0t-format` for callers that want it directly instead of
 * going through this adapter.
 */

import type { KnotByteSource, KnotMetadata } from './types';
import {
  GGML_TYPE_NAMES,
  GGML_TYPE_OTHER,
  KN0T_MAGIC,
  readKn0tHeader as readKn0tHeaderBytes,
  readKn0tTocRecord,
  type Kn0tHeader,
} from '@a0n/bitwise/kn0t-format';

const KNOT_MAGIC = 'KNOT';
const HEADER_PREFIX_BYTES = 10; // magic(4) + version(2) + meta_len(4)

export interface KnotHeader {
  version: number;
  metadataLength: number;
  /** Parsed metadata JSON. */
  metadata: KnotMetadata;
  /** Raw metadata JSON string, useful when forwarding straight to WASM. */
  metadataJson: string;
  /** Absolute byte offset where the tensor payload begins. */
  payloadOffset: number;
  /** Codec marker byte (one byte sitting between padded metadata and payload). */
  codecMarker: number;
}

/**
 * Compute the payload offset given the metadata length, per the format spec:
 *   ((10 + meta_len + 3) & ~3) + 1
 *
 * The bitwise & ~3 rounds up to the next 4-byte boundary; the trailing + 1
 * skips the single codec marker byte.
 */
export function computePayloadOffset(metadataLength: number): number {
  const padded = (HEADER_PREFIX_BYTES + metadataLength + 3) & ~3;
  return padded + 1;
}

/**
 * Adapt a parsed `.kn0t` header into the same `KnotHeader` shape
 * `.knot`'s JSON-TOC path produces, so downstream consumers don't need to
 * know which format backed the file.
 *
 * Two semantics to get right, both load-bearing:
 * - `.kn0t`'s TOC records store *absolute* file offsets; `.knot`'s JSON
 *   TOC stores offsets *relative to* `payloadOffset` (every consumer does
 *   `payloadOffset + info.offset`, e.g. `KnotModelStorage.getChunk`) — so
 *   this subtracts `payloadOffset` back out per record.
 * - a `type_tag` of `GGML_TYPE_OTHER` means the real type name lives in
 *   `metadata.type_overrides[name]`, not in the GGML table.
 */
interface RawTensorEntry {
  offset: number;
  length: number;
  /** GGUF-style quant tag, e.g. "Q4_K", "F16", "F32" — matches the field
   * name `JSON.parse` actually produces for `.knot`'s real TOC (`.type`,
   * not `KnotMetadata`'s declared `TensorInfo.tensor_type`; that mismatch
   * predates this adapter and nothing downstream reads `.tensor_type`, so
   * this matches the real v1 runtime shape rather than the aspirational
   * declared one). */
  type: string;
}

function kn0tHeaderToKnotHeader(kn0t: Kn0tHeader): KnotHeader {
  const typeOverrides = (kn0t.metadata.type_overrides ?? {}) as Record<string, string>;
  const tensors: Record<string, RawTensorEntry> = {};
  for (let i = 0; i < kn0t.recordCount; i += 1) {
    const rec = readKn0tTocRecord(kn0t, i);
    const type =
      rec.typeTag === GGML_TYPE_OTHER ? typeOverrides[rec.name] : GGML_TYPE_NAMES[rec.typeTag];
    tensors[rec.name] = {
      offset: rec.offset - kn0t.payloadOffset,
      length: rec.length,
      type: type ?? 'UNKNOWN',
    };
  }

  const { type_overrides: _typeOverrides, ...restMetadata } = kn0t.metadata as Record<
    string,
    unknown
  >;
  const metadata = { ...restMetadata, tensors } as unknown as KnotMetadata;

  return {
    version: kn0t.version,
    metadataLength: kn0t.metaLen,
    metadata,
    metadataJson: JSON.stringify(metadata),
    payloadOffset: kn0t.payloadOffset,
    codecMarker: kn0t.codecMarker,
  };
}

/**
 * Read header from an in-memory window starting at byte 0 of the knot.
 * Handles both `.knot` and `.kn0t` — see the adapter note above.
 *
 * Throws if the magic is wrong or the window is shorter than what the header
 * claims it needs. Caller is responsible for re-fetching with a larger
 * window if `metadataLength + payload_offset_overhead` exceeds what was read.
 */
export function parseKnotHeaderFromBytes(window: Uint8Array): KnotHeader {
  if (window.length < 4) {
    throw new Error(`knot header: window too small (${window.length} bytes, need at least 4)`);
  }
  const magic4 = new TextDecoder('utf-8').decode(window.subarray(0, 4));
  if (magic4 === KN0T_MAGIC) {
    return kn0tHeaderToKnotHeader(readKn0tHeaderBytes(window));
  }

  if (window.length < HEADER_PREFIX_BYTES) {
    throw new Error(
      `knot header: window too small (${window.length} bytes, need at least ${HEADER_PREFIX_BYTES})`,
    );
  }

  const magic = new TextDecoder('utf-8').decode(window.subarray(0, 4));
  if (magic !== KNOT_MAGIC) {
    throw new Error(
      `knot header: bad magic ${JSON.stringify(magic)} (expected "KNOT" or "${KN0T_MAGIC}")`,
    );
  }

  // DataView constructed against the same underlying buffer but offset into
  // the view's window so byteOffset arithmetic works for sub-arrays too.
  const view = new DataView(window.buffer, window.byteOffset, window.byteLength);
  const version = view.getUint16(4, /* littleEndian */ true);
  const metadataLength = view.getUint32(6, /* littleEndian */ true);

  const payloadOffset = computePayloadOffset(metadataLength);
  const metadataEnd = HEADER_PREFIX_BYTES + metadataLength;

  if (window.length < payloadOffset) {
    throw new Error(
      `knot header: window too small for metadata + codec marker ` +
        `(have ${window.length}, need ${payloadOffset})`,
    );
  }

  const metadataBytes = window.subarray(HEADER_PREFIX_BYTES, metadataEnd);
  const metadataJson = new TextDecoder('utf-8').decode(metadataBytes);
  let metadata: KnotMetadata;
  try {
    metadata = JSON.parse(metadataJson) as KnotMetadata;
  } catch (err) {
    throw new Error(
      `knot header: metadata JSON parse failed: ${(err as Error).message}`,
    );
  }
  if (!metadata || typeof metadata !== 'object' || !metadata.tensors) {
    throw new Error('knot header: metadata missing required `tensors` field');
  }

  // Codec marker sits one byte before payloadOffset.
  const codecMarker = window[payloadOffset - 1] ?? 0;

  return {
    version,
    metadataLength,
    metadata,
    metadataJson,
    payloadOffset,
    codecMarker,
  };
}

/**
 * Fetch and parse the knot header (either format). If the initial probe is
 * too small to capture the declared metadata, issues a second range
 * request that does.
 *
 * The two formats need different widen strategies: `.knot`'s exact byte
 * need is computable up front from `meta_len` alone (`computePayloadOffset`).
 * `.kn0t`'s exact need additionally depends on the string table's total
 * length, which isn't knowable without walking it — so on a `.kn0t` miss
 * this widens once to a large fixed ceiling rather than computing an exact
 * size. In practice this is moot: real callers pass multi-MB
 * `initialFetchBytes` (e.g. `HostShim`'s 4 MiB default), which already
 * covers any real model's TOC + string table + metadata (kilobytes), so
 * the widen path only fires for pathologically small initial probes.
 */
export async function readKnotHeader(
  source: KnotByteSource,
  initialFetchBytes: number,
): Promise<KnotHeader> {
  if (initialFetchBytes < HEADER_PREFIX_BYTES) {
    throw new Error(
      `readKnotHeader: initialFetchBytes ${initialFetchBytes} < required prefix ${HEADER_PREFIX_BYTES}`,
    );
  }
  const probe = await source.rangeGet(0, initialFetchBytes - 1);
  if (probe.length < HEADER_PREFIX_BYTES) {
    throw new Error(
      `readKnotHeader: source returned ${probe.length} bytes, need >= ${HEADER_PREFIX_BYTES}`,
    );
  }

  const magic4 = new TextDecoder('utf-8').decode(probe.subarray(0, 4));
  if (magic4 === KN0T_MAGIC) {
    try {
      return parseKnotHeaderFromBytes(probe);
    } catch {
      // Probe was too small for the full TOC + string table + metadata —
      // widen once to a large fixed ceiling (see doc comment above) and
      // let a genuine parse failure propagate from this second attempt.
      const widenBytes = Math.max(initialFetchBytes * 4, 16 * 1024 * 1024);
      const widened = await source.rangeGet(0, widenBytes - 1);
      return parseKnotHeaderFromBytes(widened);
    }
  }

  // Read the prefix to learn metadata length, then decide if we need a wider window.
  const view = new DataView(probe.buffer, probe.byteOffset, probe.byteLength);
  const declaredMetaLen = view.getUint32(6, true);
  const needed = computePayloadOffset(declaredMetaLen);

  if (probe.length >= needed) {
    return parseKnotHeaderFromBytes(probe);
  }

  // Re-fetch a window that covers the full metadata + codec marker.
  const widened = await source.rangeGet(0, needed - 1);
  if (widened.length < needed) {
    throw new Error(
      `readKnotHeader: widened fetch returned ${widened.length}, need ${needed}`,
    );
  }
  return parseKnotHeaderFromBytes(widened);
}
