// @a0n/.ignored_gnosis/distributed-inference-host/src/liquid-memory-session.ts
import {
  Pipeline,
  prefixHashU64,
  type PipelinePrefillTelemetry,
  type PipelineTokenEvent,
  type PipelinePrefillProgress,
  type PrefillResult,
} from './pipeline.ts';
import {
  computeMemoryPressure,
  createLiquidMemoryNode,
  decideLiquidFlow,
  routeItem,
  ventItem,
  type LiquidItem,
  type LiquidMemoryNode,
} from '../../src/liquid-memory-topology.ts';

type LiquidSessionStatus =
  | 'disabled'
  | 'hit'
  | 'miss'
  | 'capture_failed';

export interface LiquidMemoryTelemetry {
  status: LiquidSessionStatus;
  prefixHash?: string;
  prefixLength?: number;
  sessionId?: string;
  node?: 'hot' | 'warm';
  restoreMs?: number;
  saveMs?: number;
  kvSaved?: boolean;
  pressure: {
    hot: number;
    warm: number;
  };
  items: {
    hot: number;
    warm: number;
  };
  ventCount: number;
  routeCount: number;
  missReason?: string;
  prefill?: PipelinePrefillTelemetry;
}

export interface LiquidMemoryGenerationResult {
  tokens: Uint32Array;
  telemetry: LiquidMemoryTelemetry;
}

export interface LiquidMemorySessionManagerConfig {
  pipeline: Pipeline;
  enabled?: boolean;
  hotCapacity?: number;
  warmCapacity?: number;
  now?: () => number;
}

export interface LiquidMemoryGenerateArgs {
  model: string;
  promptTokens: Uint32Array;
  maxNewTokens: number;
  onPrefillProgress?: (progress: PipelinePrefillProgress) => void;
  onToken?: (event: PipelineTokenEvent) => void;
  onTelemetry?: (telemetry: LiquidMemoryTelemetry) => void;
}

interface LiquidSessionRecord {
  model: string;
  prefixHash: string;
  prefixLength: number;
  sessionId: string;
  prefill: PrefillResult;
  createdAt: number;
  lastAccessedAt: number;
  uses: number;
  kvSaved: boolean;
  prefillTelemetry: PipelinePrefillTelemetry;
}

function copyFloat32(value: Float32Array): Float32Array {
  const out = new Float32Array(value.length);
  out.set(value);
  return out;
}

function retainPrefill(prefill: PrefillResult): PrefillResult {
  const retained: PrefillResult = {
    logits: copyFloat32(prefill.logits),
    stationCallCount: prefill.stationCallCount,
    poisoned: prefill.poisoned,
    poisonChunk: prefill.poisonChunk,
  };
  if (prefill.tailResidual) {
    retained.tailResidual = copyFloat32(prefill.tailResidual);
  }
  if (prefill.bizarroDispatches !== undefined) {
    retained.bizarroDispatches = prefill.bizarroDispatches;
  }
  if (prefill.bizarroRetirements !== undefined) {
    retained.bizarroRetirements = prefill.bizarroRetirements;
  }
  return retained;
}

function clonePrefill(prefill: PrefillResult): PrefillResult {
  return retainPrefill(prefill);
}

function sessionKey(model: string, prefixHash: string, prefixLength: number): string {
  return `${model.trim().toLowerCase()}:${prefixLength}:${prefixHash}`;
}

function stationSessionSegment(value: string): string {
  const segment = value
    .trim()
    .replace(/[^A-Za-z0-9_-]+/g, '_')
    .replace(/^_+|_+$/g, '');
  return segment || 'x';
}

/**
 * Handles the gnosis liquid Memory Station Session Id workflow.
 */
export function liquidMemoryStationSessionId(
  model: string,
  prefixHash: string,
  prefixLength: number,
): string {
  const modelSegment = stationSessionSegment(model.toLowerCase()).slice(0, 48);
  const prefixSegment = stationSessionSegment(prefixHash).slice(0, 32);
  const lengthSegment = String(Math.max(0, Math.floor(prefixLength)));
  const id = `liquid_${modelSegment}_${lengthSegment}_${prefixSegment}`;
  if (id.length <= 128) return id;
  return `liquid_${modelSegment.slice(0, 32)}_${lengthSegment}_${prefixSegment}`;
}

function densityFor(record: LiquidSessionRecord): number {
  const useBoost = Math.min(0.35, record.uses * 0.05);
  const lengthBoost = Math.min(0.25, record.prefixLength / 512);
  return Math.min(1, 0.45 + useBoost + lengthBoost);
}

function heatDebtFor(record: LiquidSessionRecord): number {
  return Math.max(1, Math.ceil(record.prefixLength / 8));
}

export class LiquidMemorySessionManager {
  private readonly pipeline: Pipeline;
  private readonly enabled: boolean;
  private readonly now: () => number;
  private readonly hot: LiquidMemoryNode<LiquidSessionRecord>;
  private readonly warm: LiquidMemoryNode<LiquidSessionRecord>;
  private ventCount = 0;
  private routeCount = 0;
  /**
   * Runtime circuit breaker. The boot capability probe treats any non-404 from
   * `/kv-session/save` (including a 500) as "supported", and a `prefix_len=0`
   * probe can't catch a station that only fails on REAL prefixes. So a station
   * that can't actually save KV would otherwise pay the `capture_failed`
   * re-prefill penalty (~0.58 tok/s) on EVERY request. After
   * `circuitThreshold` consecutive real-save failures we open the breaker and
   * route to the native fast path (~7 tok/s) for the rest of the process.
   */
  private consecutiveCaptureFailures = 0;
  private circuitOpen = false;
  private readonly circuitThreshold = 3;

  constructor(config: LiquidMemorySessionManagerConfig) {
    this.pipeline = config.pipeline;
    this.enabled = config.enabled ?? true;
    this.now = config.now ?? Date.now;
    const hotCapacity = Math.max(1, Math.floor(config.hotCapacity ?? 8));
    const warmCapacity = Math.max(1, Math.floor(config.warmCapacity ?? 24));
    let hot!: LiquidMemoryNode<LiquidSessionRecord>;
    let warm!: LiquidMemoryNode<LiquidSessionRecord>;
    hot = createLiquidMemoryNode<LiquidSessionRecord>('hot', hotCapacity, hotCapacity, () => [warm]);
    warm = createLiquidMemoryNode<LiquidSessionRecord>('warm', warmCapacity, warmCapacity, () => [hot]);
    this.hot = hot;
    this.warm = warm;
    this.syncBudget(this.hot);
    this.syncBudget(this.warm);
  }

  snapshot(
    status: LiquidSessionStatus,
    fields: Partial<Omit<LiquidMemoryTelemetry, 'status' | 'pressure' | 'items' | 'ventCount' | 'routeCount'>> = {},
  ): LiquidMemoryTelemetry {
    this.syncBudget(this.hot);
    this.syncBudget(this.warm);
    return {
      status,
      ...fields,
      pressure: {
        hot: computeMemoryPressure(this.hot),
        warm: computeMemoryPressure(this.warm),
      },
      items: {
        hot: this.hot.items.size,
        warm: this.warm.items.size,
      },
      ventCount: this.ventCount,
      routeCount: this.routeCount,
    };
  }

  async generate(args: LiquidMemoryGenerateArgs): Promise<LiquidMemoryGenerationResult> {
    if (!this.enabled || this.circuitOpen) {
      const tokens = await this.pipeline.generate({
        promptTokens: args.promptTokens,
        maxNewTokens: args.maxNewTokens,
        onPrefillProgress: args.onPrefillProgress,
        onToken: args.onToken,
      });
      const telemetry = this.snapshot('disabled');
      args.onTelemetry?.(telemetry);
      return { tokens, telemetry };
    }

    const prefixLength = args.promptTokens.length;
    const prefixHash = await prefixHashU64(args.promptTokens);
    const existing = this.find(args.model, prefixHash, prefixLength);
    if (existing) {
      const restoreStartedAt = this.now();
      const restored = existing.item.value.kvSaved
        ? await this.pipeline
            .loadKvSession(existing.item.value.sessionId)
            .catch(() => null)
        : args.maxNewTokens <= 1
          ? { ok: true, prefixLength }
          : null;
      const restoreMs = Math.max(0, this.now() - restoreStartedAt);
      if (restored?.prefixLength === prefixLength) {
        existing.item.value.uses += 1;
        existing.item.value.lastAccessedAt = this.now();
        existing.item.informationDensity = densityFor(existing.item.value);
        this.promote(existing);
        const telemetry = this.snapshot('hit', {
          prefixHash,
          prefixLength,
          sessionId: existing.item.value.sessionId,
          node: existing.nodeId,
          restoreMs,
          kvSaved: existing.item.value.kvSaved,
          prefill: existing.item.value.prefillTelemetry,
        });
        args.onTelemetry?.(telemetry);
        const tokens = await this.pipeline.generateFromPrefill({
          promptTokens: args.promptTokens,
          prefill: clonePrefill(existing.item.value.prefill),
          maxNewTokens: args.maxNewTokens,
          onToken: args.onToken,
        });
        return { tokens, telemetry };
      }
      if (existing.item.value.kvSaved || args.maxNewTokens > 1) {
        this.deleteItem(existing.node, existing.item.key);
      }
    }

    const prepared = await this.pipeline.preparePrefill({
      promptTokens: args.promptTokens,
      onPrefillProgress: args.onPrefillProgress,
    });
    const sessionId = liquidMemoryStationSessionId(args.model, prefixHash, prefixLength);
    const saveStartedAt = this.now();
    // Capture the save failure reason instead of swallowing it. saveKvSession
    // returns false on HTTP 404 (route missing) and throws on other non-2xx
    // (e.g. 500 with the station's "kv session save panicked: <reason>" body).
    // Surfacing this in missReason makes a broken save diagnosable from the
    // `_gnosis` telemetry alone, without station-side log forensics.
    let saveError: string | undefined;
    const kvSaved = await this.pipeline
      .saveKvSession(sessionId, prefixLength)
      .catch((e: unknown) => {
        saveError = e instanceof Error ? e.message : String(e);
        return false;
      });
    const saveMs = Math.max(0, this.now() - saveStartedAt);

    const status: LiquidSessionStatus = kvSaved ? 'miss' : 'capture_failed';
    // Circuit breaker: trip after repeated real-save failures so we stop paying
    // the capture_failed re-prefill penalty on every subsequent request.
    if (kvSaved) {
      this.consecutiveCaptureFailures = 0;
    } else if (++this.consecutiveCaptureFailures >= this.circuitThreshold) {
      this.circuitOpen = true;
      console.warn(
        `[liquid] circuit OPEN after ${this.consecutiveCaptureFailures} consecutive ` +
          `kv-save failures — disabling liquid for this process (native fast path). ` +
          `Last reason: ${saveError ?? 'HTTP 404 (route missing)'}`,
      );
    }
    if (!prepared.prefill.poisoned) {
      this.remember({
        model: args.model,
        prefixHash,
        prefixLength,
        sessionId,
        prefill: retainPrefill(prepared.prefill),
        createdAt: this.now(),
        lastAccessedAt: this.now(),
        uses: 0,
        kvSaved,
        prefillTelemetry: prepared.telemetry,
      });
    }

    const telemetry = this.snapshot(status, {
      prefixHash,
      prefixLength,
      sessionId,
      saveMs,
      kvSaved,
      prefill: prepared.telemetry,
      // saveError set ⇒ non-2xx (e.g. 500 panic body); undefined ⇒ HTTP 404
      // (route missing on a stale/incompatible station binary).
      ...(kvSaved
        ? {}
        : {
            missReason:
              saveError ??
              'station kv session save failed (HTTP 404 — /kv-session/save route missing)',
          }),
    });
    args.onTelemetry?.(telemetry);
    const tokens = await this.pipeline.generateFromPrefill({
      promptTokens: args.promptTokens,
      prefill: prepared.prefill,
      maxNewTokens: args.maxNewTokens,
      onToken: args.onToken,
    });
    return { tokens, telemetry };
  }

  private remember(record: LiquidSessionRecord): void {
    const key = sessionKey(record.model, record.prefixHash, record.prefixLength);
    const existing = this.find(record.model, record.prefixHash, record.prefixLength);
    if (existing) {
      existing.item.value = record;
      existing.item.informationDensity = densityFor(record);
      existing.item.heatDebt = heatDebtFor(record);
      existing.item.insertedAt = record.createdAt;
      this.promote(existing);
      return;
    }

    const item: LiquidItem<LiquidSessionRecord> = {
      key,
      value: record,
      insertedAt: record.createdAt,
      informationDensity: densityFor(record),
      heatDebt: heatDebtFor(record),
    };
    this.syncBudget(this.hot);
    const decision = decideLiquidFlow(this.hot, item, this.now());
    if (this.hot.items.size >= this.hot.budget.capacity) {
      if (decision.action === 'route') {
        this.routeLowestHotToWarm();
      } else {
        this.ventLowest(this.hot);
      }
    }
    this.hot.items.set(key, item);
    this.syncBudget(this.hot);
  }

  private find(
    model: string,
    prefixHash: string,
    prefixLength: number,
  ):
    | {
        node: LiquidMemoryNode<LiquidSessionRecord>;
        nodeId: 'hot' | 'warm';
        item: LiquidItem<LiquidSessionRecord>;
      }
    | null {
    const key = sessionKey(model, prefixHash, prefixLength);
    const hotItem = this.hot.items.get(key);
    if (hotItem) return { node: this.hot, nodeId: 'hot', item: hotItem };
    const warmItem = this.warm.items.get(key);
    if (warmItem) return { node: this.warm, nodeId: 'warm', item: warmItem };
    return null;
  }

  private promote(entry: {
    node: LiquidMemoryNode<LiquidSessionRecord>;
    nodeId: 'hot' | 'warm';
    item: LiquidItem<LiquidSessionRecord>;
  }): void {
    if (entry.nodeId === 'hot') return;
    if (this.hot.items.size >= this.hot.budget.capacity) {
      this.routeLowestHotToWarm();
    }
    entry.node.items.delete(entry.item.key);
    this.hot.items.set(entry.item.key, entry.item);
    this.routeCount += 1;
    this.syncBudget(entry.node);
    this.syncBudget(this.hot);
  }

  private routeLowestHotToWarm(): void {
    const candidate = this.lowestDensity(this.hot);
    if (!candidate) return;
    if (this.warm.items.size >= this.warm.budget.capacity) {
      this.ventLowest(this.warm);
    }
    this.hot.items.set(candidate.key, candidate);
    if (routeItem(this.hot, candidate.key, this.warm)) {
      this.routeCount += 1;
    }
    this.syncBudget(this.hot);
    this.syncBudget(this.warm);
  }

  private ventLowest(node: LiquidMemoryNode<LiquidSessionRecord>): void {
    const candidate = this.lowestDensity(node);
    if (!candidate) return;
    const heat = ventItem(node, candidate.key);
    if (heat > 0 || !node.items.has(candidate.key)) {
      this.ventCount += 1;
    }
    this.syncBudget(node);
  }

  private deleteItem(node: LiquidMemoryNode<LiquidSessionRecord>, key: string): void {
    node.items.delete(key);
    this.syncBudget(node);
  }

  private lowestDensity(
    node: LiquidMemoryNode<LiquidSessionRecord>,
  ): LiquidItem<LiquidSessionRecord> | null {
    let candidate: LiquidItem<LiquidSessionRecord> | null = null;
    for (const item of node.items.values()) {
      if (
        candidate === null ||
        item.informationDensity < candidate.informationDensity ||
        (item.informationDensity === candidate.informationDensity &&
          item.value.lastAccessedAt < candidate.value.lastAccessedAt)
      ) {
        candidate = item;
      }
    }
    return candidate;
  }

  private syncBudget(node: LiquidMemoryNode<LiquidSessionRecord>): void {
    node.budget.held = node.items.size;
  }
}

/**
 * Handles the gnosis liquid Memory Telemetry Headers workflow.
 */
export function liquidMemoryTelemetryHeaders(
  telemetry: LiquidMemoryTelemetry | undefined,
): Record<string, string> {
  if (!telemetry) return {};
  const headers: Record<string, string> = {
    'X-Thoth-Liquid-Memory': telemetry.status,
    'X-Thoth-Liquid-Pressure-Hot': telemetry.pressure.hot.toFixed(1),
    'X-Thoth-Liquid-Pressure-Warm': telemetry.pressure.warm.toFixed(1),
    'X-Thoth-Liquid-Vents': String(telemetry.ventCount),
    'X-Thoth-Liquid-Routes': String(telemetry.routeCount),
  };
  if (telemetry.prefixHash !== undefined) {
    headers['X-Thoth-Liquid-Prefix-Hash'] = telemetry.prefixHash;
  }
  if (telemetry.prefixLength !== undefined) {
    headers['X-Thoth-Liquid-Prefix-Tokens'] = String(telemetry.prefixLength);
  }
  if (telemetry.sessionId !== undefined) {
    headers['X-Thoth-Liquid-Session'] = telemetry.sessionId;
  }
  if (telemetry.restoreMs !== undefined) {
    headers['X-Thoth-Liquid-Restore-Ms'] = String(telemetry.restoreMs);
  }
  if (telemetry.saveMs !== undefined) {
    headers['X-Thoth-Liquid-Save-Ms'] = String(telemetry.saveMs);
  }
  if (telemetry.kvSaved !== undefined) {
    headers['X-Thoth-Liquid-KV-Saved'] = telemetry.kvSaved ? 'true' : 'false';
  }
  if (telemetry.prefill) {
    headers['X-Thoth-Liquid-Prefill-Source'] = telemetry.prefill.source;
    headers['X-Thoth-Liquid-Replay-Ms'] = String(telemetry.prefill.replayMs);
    headers['X-Thoth-Liquid-Prefill-Ms'] = String(telemetry.prefill.prefillMs);
  }
  if (telemetry.missReason) {
    headers['X-Thoth-Liquid-Miss-Reason'] = telemetry.missReason;
  }
  return headers;
}
