// @a0n/.ignored_gnosis/distributed-inference-host/src/lru.ts
/**
 * Tiny LRU keyed by transformer layer index.
 *
 * Map preserves insertion order, so the oldest key is the LRU victim. We
 * delete-then-set on touch to bump a key to the most-recently-used slot.
 *
 * Capacity is enforced on `recordTouch`; if the map exceeds capacity, the
 * oldest key (other than the just-touched key) is returned as `evicted`.
 */

export class LayerLRU {
  private readonly order = new Map<number, true>();

  constructor(private readonly capacity: number) {
    if (!Number.isInteger(capacity) || capacity < 1) {
      throw new Error(`LayerLRU: capacity must be a positive integer, got ${capacity}`);
    }
  }

  /**
   * Record a touch on `layerIdx`. Returns the layer index that was evicted
   * (if any) to make room. Returns null when no eviction occurred.
   */
  recordTouch(layerIdx: number): number | null {
    if (this.order.has(layerIdx)) {
      // Bump to MRU position.
      this.order.delete(layerIdx);
      this.order.set(layerIdx, true);
      return null;
    }
    this.order.set(layerIdx, true);
    if (this.order.size <= this.capacity) {
      return null;
    }
    // Evict the oldest key (first iterator value).
    const oldestIter = this.order.keys().next();
    if (oldestIter.done) {
      return null;
    }
    const victim = oldestIter.value;
    this.order.delete(victim);
    return victim;
  }

  /** Currently resident layers, ordered LRU -> MRU. */
  residentLayers(): number[] {
    return Array.from(this.order.keys());
  }

  has(layerIdx: number): boolean {
    return this.order.has(layerIdx);
  }

  get size(): number {
    return this.order.size;
  }

  clear(): void {
    this.order.clear();
  }
}
