// @a0n/.ignored_gnosis/distributed-inference-host/src/microphone-fingerprint.ts
/**
 * # Microphone Fingerprint — device provenance, spoof / deepfake refutation
 *
 * Mirrors `Gnosis.Mesh.MicrophoneFingerprint` from `open-source/gnosis-math`,
 * the Rustic Church carrier for the microphone-fingerprint idea framed
 * *defensively* as a provenance / spoofed-device detector. Every physical
 * microphone has a frequency-response signature — a per-band gain vector fixed
 * by its diaphragm, capsule, and analog front end. Two recordings made on the
 * *same* device carry the same fingerprint; recordings from *different* devices
 * are separable by that fingerprint. A "this clip came from that device"
 * provenance claim is exactly what an attacker wants to forge (a deepfake or a
 * spoofed-device impersonation), so the Skeptic admits a same-device claim only
 * when the fingerprints actually match AND the fingerprint is non-trivial — an
 * all-zero gain vector carries no device identity and could come from anywhere.
 *
 * ## What is (and is not) carried
 *
 * This module is the runtime promotion check for a same-device provenance
 * observation. It carries ONLY the finite contract the Lean file proves:
 *
 *  - the saturating per-band gain metric (`adist`);
 *  - the fingerprint distance as the summed pointwise `adist`, with a large
 *    sentinel on a length mismatch (`fpDistance` / `mismatchPenalty`);
 *  - the non-triviality gate as the source-only-rejection analog: an all-zero
 *    fingerprint has zero total gain and carries no identity
 *    (`totalGain` / `nonTrivial`);
 *  - same-device admission: match within tolerance AND non-trivial
 *    (`admitSameDevice`), with the distinguishability hard negative folded into
 *    the withheld outcome.
 *
 * Real frequency-response estimation, sweep / chirp calibration, room and codec
 * coloration, anti-spoof liveness, and the actual statistical separation of
 * device populations stay runtime/empirical obligations OUTSIDE this
 * certificate — they are not theorems and are not asserted here.
 *
 * The promotion check folds the contract into the same epistemic-outcome
 * vocabulary the sibling mesh carriers already speak
 * (`claim_admitted` / `observation_only` / `claim_withheld`), so a same-device
 * provenance claim cannot be displayed until it has cleared the Lean-side
 * match/non-triviality gate.
 *
 * ## Lean source-of-truth
 *
 * @see open-source/gnosis-math/Gnosis/Mesh/MicrophoneFingerprint.lean
 * @see open-source/gnosis-math/Gnosis/Mesh/ElectricalNetworkFrequency.lean (mesh sibling)
 */

/* ------------------------------------------------------------------ */
/*  Per-band distance and fingerprints                                */
/* ------------------------------------------------------------------ */

/**
 * A microphone fingerprint is a list of per-band gains over a fixed window.
 * Gains are non-negative integers (saturating `Nat` arithmetic), matching the
 * Lean `Fingerprint`.
 */
export type Fingerprint = readonly number[];

/** Absolute difference on `Nat` via saturating subtraction (`|a - b|`). The
 *  per-band gain residual metric, shared with the mesh corridor carriers.
 *  Mirrors `MicrophoneFingerprint.adist`. */
export function adist(a: number, b: number): number {
  return Math.max(0, a - b) + Math.max(0, b - a);
}

/**
 * Sentinel returned on a length mismatch: large enough that no sane tolerance
 * admits it, so differently-windowed fingerprints are treated as non-matching
 * (different windowing => not the same recording geometry => refuse the
 * same-device claim). Mirrors `MicrophoneFingerprint.mismatchPenalty`.
 */
export const mismatchPenalty = 1000000;

/**
 * The fingerprint distance: the sum of pointwise `adist` between aligned bands.
 * On a length mismatch it returns `mismatchPenalty` rather than attempting a
 * partial alignment. Mirrors `MicrophoneFingerprint.fpDistance`.
 */
export function fpDistance(x: Fingerprint, y: Fingerprint): number {
  if (x.length !== y.length) {
    return mismatchPenalty;
  }
  let sum = 0;
  for (let i = 0; i < x.length; i++) {
    sum += adist(x[i], y[i]);
  }
  return sum;
}

/* ------------------------------------------------------------------ */
/*  Non-triviality gate (source-only rejection analog)                */
/* ------------------------------------------------------------------ */

/**
 * Total gain of a fingerprint: the summed per-band gains. An all-zero
 * fingerprint has zero total gain and carries no device identity.
 * Mirrors `MicrophoneFingerprint.totalGain`.
 */
export function totalGain(fp: Fingerprint): number {
  let sum = 0;
  for (const g of fp) {
    sum += g;
  }
  return sum;
}

/**
 * A fingerprint is non-trivial when it carries some gain (some band is
 * nonzero). An all-zero fingerprint fails this gate and carries no identity.
 * Mirrors `MicrophoneFingerprint.NonTrivial` / `nonTrivialB`.
 */
export function nonTrivial(fp: Fingerprint): boolean {
  return totalGain(fp) > 0;
}

/* ------------------------------------------------------------------ */
/*  Runtime promotion check (epistemic outcome)                       */
/* ------------------------------------------------------------------ */

/**
 * Epistemic outcome of a same-device provenance observation. Identical
 * vocabulary to the sibling mesh carriers (`./cross-node-coincidence`,
 * `./electrical-network-frequency`), kept as a local literal union so this
 * certificate has no import coupling:
 *
 *  - `claim_admitted`    — the two fingerprints match within tolerance AND the
 *                          fingerprint is non-trivial, so the same-device
 *                          provenance claim may be promoted.
 *  - `observation_only`  — reserved for the trivial fingerprint that matches
 *                          itself: the gate sees a (vacuous) match but the
 *                          fingerprint carries no identity, so it is reported,
 *                          never promoted. (Used here for the all-zero case so a
 *                          trivial self-match is not silently withheld.)
 *  - `claim_withheld`    — the fingerprints are separated by more than tolerance
 *                          (different device / not same), so there is no
 *                          same-device claim to make.
 */
export type MicFingerprintEpistemicOutcome =
  | 'claim_admitted'
  | 'observation_only'
  | 'claim_withheld';

/** The reason a same-device claim was held back from promotion. */
export type MicFingerprintLimitationClass =
  | 'no_device_identity'
  | 'different_device'
  | 'none';

export interface MicFingerprintAdmission {
  readonly outcome: MicFingerprintEpistemicOutcome;
  readonly limitation: MicFingerprintLimitationClass;
  /** Whether the two fingerprints cleared the distance tolerance. */
  readonly matched: boolean;
  /** Whether the first fingerprint carries device identity (non-trivial). */
  readonly nonTrivial: boolean;
  /** The measured fingerprint distance (the displayable separation). */
  readonly distance: number;
  readonly summary: string;
}

export interface MicFingerprintAdmitOptions {
  /** Tolerance for the pointwise fingerprint distance match. */
  readonly tol: number;
}

/**
 * The promotion check Monster Mesh runs before any UI surface displays a
 * "this clip came from that device" provenance claim. It folds the Lean
 * contract into an epistemic outcome:
 *
 *  - the non-triviality gate decides whether there is any device identity at all
 *    — an all-zero fingerprint carries none, so even a self-match is withheld /
 *    reported but never promoted (the source-only rejection analog, mirroring
 *    `trivial_fingerprint_rejected`);
 *  - the match gate decides whether the two recordings actually share the same
 *    fingerprint — a separation beyond tolerance is a different device and is
 *    withheld (mirroring `distinct_device_refuted` / `different_device_refuted`);
 *  - only a non-trivial fingerprint matching within tolerance is admitted as a
 *    same-device claim (mirroring `same_device_admitted`).
 *
 * Note the trivial-fingerprint case: even though `fpDistance fp fp = 0` would
 * clear any tolerance, the Lean carrier refuses to admit it (a vacuous identity
 * match). Here that lands as `observation_only` (a present-but-uncertifiable
 * self-match) when the trivial fingerprint matches, and `claim_withheld` when it
 * is even separated.
 *
 * @param x    the candidate recording's fingerprint (the one whose identity is
 *             being asserted)
 * @param y    the reference recording's fingerprint
 * @param opts the distance tolerance
 */
export function admitSameDevice(
  x: Fingerprint,
  y: Fingerprint,
  opts: MicFingerprintAdmitOptions,
): MicFingerprintAdmission {
  const distance = fpDistance(x, y);
  const matched = distance <= opts.tol;
  const identity = nonTrivial(x);

  // Non-triviality gate: an all-zero fingerprint carries no device identity, so
  // there is no same-device claim to promote regardless of how close it sits.
  if (!identity) {
    if (matched) {
      return {
        outcome: 'observation_only',
        limitation: 'no_device_identity',
        matched,
        nonTrivial: identity,
        distance,
        summary:
          'trivial (all-zero) fingerprint matches but carries no device identity — vacuous self-match, not certified',
      };
    }
    return {
      outcome: 'claim_withheld',
      limitation: 'no_device_identity',
      matched,
      nonTrivial: identity,
      distance,
      summary:
        'trivial (all-zero) fingerprint carries no device identity — nothing to claim',
    };
  }

  // Match gate: a non-trivial fingerprint separated by more than tolerance is a
  // different device, so there is no same-device claim.
  if (!matched) {
    return {
      outcome: 'claim_withheld',
      limitation: 'different_device',
      matched,
      nonTrivial: identity,
      distance,
      summary: `fingerprints separated by ${distance} > tol ${opts.tol} — different device, not the same`,
    };
  }

  return {
    outcome: 'claim_admitted',
    limitation: 'none',
    matched,
    nonTrivial: identity,
    distance,
    summary: `non-trivial fingerprints match within tolerance (distance ${distance} <= ${opts.tol}) — same device certified`,
  };
}

/* ------------------------------------------------------------------ */
/*  Concrete fixtures (mirror the Lean fixtures)                      */
/* ------------------------------------------------------------------ */

/** Recording 1 of device A: its measured per-band gain vector (length 4). A
 *  genuine, non-trivial fingerprint. Mirrors `MicrophoneFingerprint.deviceA`. */
export const deviceA: Fingerprint = [12, 8, 5, 3];

/** Recording 2 of the SAME device A: a near-identical fingerprint, off by at
 *  most one quantization unit per band. `fpDistance(deviceA, deviceA2) = 0+1+1+0 = 2`,
 *  so it matches within tolerance 3. Mirrors `deviceA2`. */
export const deviceA2: Fingerprint = [12, 9, 4, 3];

/** A clearly DIFFERENT device B: its fingerprint is far from device A.
 *  `fpDistance(deviceA, deviceB) = 9+7+9+6 = 31`, well beyond tolerance 3.
 *  Mirrors `deviceB`. */
export const deviceB: Fingerprint = [3, 1, 14, 9];

/** An all-zero (trivial) fingerprint: silence / no measured response. Carries no
 *  device identity, so any same-device claim built on it is rejected.
 *  Mirrors `trivialFp`. */
export const trivialFp: Fingerprint = [0, 0, 0, 0];

/**
 * Microphone Fingerprint master witness. Bundles the boundary facts the Lean
 * theorem `mic_fingerprint_master` proves, each recomputed here from first
 * principles (never imported from Lean IR).
 *
 * Mirrors `MicrophoneFingerprint.mic_fingerprint_master`.
 */
export const MicrophoneFingerprintMaster = {
  /** A genuine re-recording matches its own fingerprint exactly (distance 0). */
  fpDistanceSelfZero: fpDistance(deviceA, deviceA) === 0,

  /** The two recordings of device A match within tolerance 3 (distance 2). */
  deviceARecordingsMatch:
    admitSameDevice(deviceA, deviceA2, { tol: 3 }).outcome === 'claim_admitted',

  /** Device B does NOT match device A within tolerance 3 (distance 31). */
  deviceBNotMatch:
    admitSameDevice(deviceA, deviceB, { tol: 3 }).outcome === 'claim_withheld',

  /** Device A's fingerprint is non-trivial: it carries genuine device identity. */
  deviceANonTrivial: nonTrivial(deviceA) === true,

  /** The all-zero fixture is rejected by the non-triviality gate. */
  trivialFpRejected: nonTrivial(trivialFp) === false,

  /** The two grids really are separated by more than tolerance 3 (distance 31),
   *  making the refutation non-vacuous. */
  devicesWellSeparated: fpDistance(deviceA, deviceB) > 3,

  /** The trivial fingerprint is NOT admitted as same-device against itself even
   *  though its self-distance is 0 (it fails the non-triviality gate). */
  trivialFingerprintRejected:
    admitSameDevice(trivialFp, trivialFp, { tol: 3 }).outcome !==
    'claim_admitted',
} as const;
