// @a0n/.ignored_gnosis/distributed-inference-host/src/openai-agentic-mode.ts
export interface AgenticModeInput {
  headers?: Record<string, string | string[] | undefined>;
  request?: {
    auto_tools?: unknown;
    execute_tools?: unknown;
    tool_choice?: unknown;
    tools?: unknown;
    mcp_url?: unknown;
  };
  envDefault?: boolean;
  envMcpUrl?: string;
}

export interface AgenticModeDecision {
  enabled: boolean;
  source: 'header' | 'body' | 'env';
  mcpUrl?: string;
}

const TRUE_VALUES = new Set(['1', 'true', 'on', 'yes', 'tools', 'auto', 'agentic']);
const FALSE_VALUES = new Set(['0', 'false', 'off', 'no', 'none', 'bare']);

function headerValue(
  headers: AgenticModeInput['headers'],
  name: string,
): string | undefined {
  if (!headers) return undefined;
  const direct = headers[name] ?? headers[name.toLowerCase()];
  if (Array.isArray(direct)) return direct[0];
  return direct;
}

/**
 * Parses the Agentic Boolean.
 */
export function parseAgenticBoolean(value: unknown): boolean | null {
  if (typeof value === 'boolean') return value;
  if (typeof value === 'number') {
    if (value === 1) return true;
    if (value === 0) return false;
  }
  if (typeof value !== 'string') return null;
  const normalized = value.trim().toLowerCase();
  if (TRUE_VALUES.has(normalized)) return true;
  if (FALSE_VALUES.has(normalized)) return false;
  return null;
}

function requestEnablesAgentic(request: AgenticModeInput['request']): boolean | null {
  if (!request) return null;

  const autoTools = parseAgenticBoolean(request.auto_tools);
  const executeTools = parseAgenticBoolean(request.execute_tools);
  if (autoTools === true || executeTools === true) return true;
  if (autoTools === false && executeTools === false) return false;

  if (Array.isArray(request.tools) && request.tools.length > 0) return true;
  if (request.tool_choice !== undefined && request.tool_choice !== null) {
    if (request.tool_choice === 'none') return false;
    return true;
  }

  return null;
}

/**
 * Resolves the Agentic Mode.
 */
export function resolveAgenticMode(
  input: AgenticModeInput,
): AgenticModeDecision {
  const headerDecision = parseAgenticBoolean(
    headerValue(input.headers, 'x-zedge-agentic'),
  );
  const headerMcpUrl = headerValue(input.headers, 'x-zedge-mcp-url');
  const bodyMcpUrl =
    typeof input.request?.mcp_url === 'string' ? input.request.mcp_url : undefined;

  if (headerDecision !== null) {
    return {
      enabled: headerDecision,
      source: 'header',
      ...(headerMcpUrl ?? bodyMcpUrl ?? input.envMcpUrl
        ? { mcpUrl: headerMcpUrl ?? bodyMcpUrl ?? input.envMcpUrl }
        : {}),
    };
  }

  const bodyDecision = requestEnablesAgentic(input.request);
  if (bodyDecision !== null) {
    return {
      enabled: bodyDecision,
      source: 'body',
      ...(bodyMcpUrl ?? headerMcpUrl ?? input.envMcpUrl
        ? { mcpUrl: bodyMcpUrl ?? headerMcpUrl ?? input.envMcpUrl }
        : {}),
    };
  }

  return {
    enabled: input.envDefault ?? false,
    source: 'env',
    ...(input.envMcpUrl ? { mcpUrl: input.envMcpUrl } : {}),
  };
}
