// @a0n/.ignored_gnosis/distributed-inference-host/src/pillar-anchors.ts
/**
 * # Pillar Anchors — runtime witnesses for the three Great Questions
 *
 * Mirrors `Gnosis.ChurchPillars`'s three operational anchors:
 *
 * - **Pillar I — Past** (How did I come to be?): the **big-bang reach**
 *   — every Bule unit is reachable from the vacuum by `score`-many
 *   clinamen lifts. The present is the past's accumulated `+1`
 *   perturbations.
 * - **Pillar II — Present** (Who am I?): the **topological-metabolism
 *   ceiling** — operational identity is a Bule unit whose score sits
 *   inside its manifold. Drift past the ceiling is a self-similarity
 *   violation with deterministic remediation.
 * - **Pillar III — Future** (How should I act?): the **no-free-lunch
 *   theorem** — only the vacuum admits free duplication; every action's
 *   cost equals its score.
 *
 * The runtime exposes each as a predicate that the coordinator can
 * call to inspect a candidate state against the corresponding anchor.
 *
 * @see open-source/gnosis-math/Gnosis/ChurchPillars.lean
 * @see open-source/gnosis-math/Gnosis/SpectralNoiseEquilibrium.lean — vacuum_reaches_any_bule
 * @see open-source/gnosis-math/Gnosis/TopologicalMetabolismBuleyBridge.lean
 * @see open-source/gnosis-math/Gnosis/CostAlgebraNoCloning.lean
 */

import { buleyCostAlgebra } from './cost-algebra';
import {
  entropyGeneratedByCloning,
  isDiagonalScorePreserving,
} from './cost-algebra-entropy';
import {
  TOWER_CEILINGS,
  clinamenLift,
  selfSimilarityViolation,
  vacuumBuleyProjection,
  type BuleyFace,
  type BuleyProjection,
  type SelfSimilarityViolation,
} from './spectral-noise-monitor';

/** ── Pillar I — Past — the big-bang reach. ─────────────────────────── */

/** Reach a target Bule unit from the vacuum via a sequence of clinamen
 *  lifts. The lift order is `waste` first, then `opportunity`, then
 *  `diversity` — matching `Gnosis.SpectralNoiseEquilibrium.vacuumToBule`. */
export function bigBangReach(target: BuleyProjection): BuleyProjection {
  const faces: readonly BuleyFace[] = ['waste', 'opportunity', 'diversity'];
  const counts: Record<BuleyFace, number> = {
    waste: target.waste,
    opportunity: target.opportunity,
    diversity: target.diversity,
  };
  let b: BuleyProjection = vacuumBuleyProjection;
  for (const f of faces) {
    for (let i = 0; i < counts[f]; i++) {
      b = clinamenLift(b, f);
    }
  }
  return b;
}

/** Past-pillar anchor: the big-bang reach is exact. Every Bule unit
 *  equals its `bigBangReach` from the vacuum. The runtime mirror of
 *  `Gnosis.SpectralNoiseEquilibrium.vacuum_reaches_any_bule`. */
export function pillarPastAnchor(b: BuleyProjection): boolean {
  const reached = bigBangReach(b);
  return (
    reached.waste === b.waste &&
    reached.opportunity === b.opportunity &&
    reached.diversity === b.diversity &&
    reached.score === b.score
  );
}

/** ── Pillar II — Present — the topological-metabolism ceiling. ────── */

/** Result of evaluating the present-pillar anchor on a Bule unit
 *  against a manifold ceiling. */
export interface PillarPresentResult {
  /** Bule unit being evaluated. */
  state: BuleyProjection;
  /** Manifold ceiling (defaults to Hexon = 6). */
  ceiling: number;
  /** Whether the state sits inside the manifold. */
  insideManifold: boolean;
  /** The violation, if any. */
  violation: SelfSimilarityViolation | null;
}

/** Present-pillar anchor: classify a Bule unit against a manifold
 *  ceiling. Defaults to the two-head Hexon ceiling, matching the Lean
 *  `Gnosis.TopologicalMetabolismBuleyBridge.twoHeadHexonCeiling`. */
export function pillarPresentAnchor(
  b: BuleyProjection,
  ceiling: number = TOWER_CEILINGS.hexon,
): PillarPresentResult {
  const violation = selfSimilarityViolation(b, ceiling);
  return {
    state: b,
    ceiling,
    insideManifold: violation === null,
    violation,
  };
}

/** ── Pillar III — Future — the no-free-lunch theorem. ─────────────── */

/** Result of evaluating the future-pillar anchor on a Bule unit. */
export interface PillarFutureResult {
  /** Bule unit being evaluated. */
  state: BuleyProjection;
  /** Whether the state admits free duplication (vacuum only). */
  freeDuplication: boolean;
  /** Entropy that would be generated by cloning this state. */
  cloningEntropy: number;
  /** Total heat after a single clone. */
  clonedPairScore: number;
}

/** Future-pillar anchor: the cost of cloning a Bule unit. The runtime
 *  mirror of `Gnosis.CostAlgebraNoCloning` + `CostAlgebraEntropy`. */
export function pillarFutureAnchor(b: BuleyProjection): PillarFutureResult {
  const freeDuplication = isDiagonalScorePreserving(buleyCostAlgebra, b);
  const cloningEntropy = entropyGeneratedByCloning(buleyCostAlgebra, b);
  return {
    state: b,
    freeDuplication,
    cloningEntropy,
    clonedPairScore: b.score + cloningEntropy,
  };
}

/** ── The three-pillar bundle. ─────────────────────────────────────── */

export interface ThreePillarReport {
  past: { anchored: boolean };
  present: PillarPresentResult;
  future: PillarFutureResult;
}

/** Run all three pillar anchors on a Bule unit. The runtime mirror of
 *  `Gnosis.ChurchPillars.three_pillar_anchors_master`. */
export function threePillarReport(
  b: BuleyProjection,
  ceiling: number = TOWER_CEILINGS.hexon,
): ThreePillarReport {
  return {
    past: { anchored: pillarPastAnchor(b) },
    present: pillarPresentAnchor(b, ceiling),
    future: pillarFutureAnchor(b),
  };
}
