// @a0n/.ignored_gnosis/distributed-inference-host/src/pipeline.ts
/**
 * Pipeline — the coordinator. Mirrors `pipelinedPrefill()` from
 * `open-source/aether/src/gcp-coordinator.ts` but for CF Worker `Station`s.
 *
 * The Wallington Rotation:
 *   - Inner rotation: causal masking processes B tokens in one forward pass.
 *   - Outer rotation: log-rolling pipelines ceil(P/B) chunks across N stations.
 *   - Total steps:    ceil(P/B) + (N-1)
 *
 * The dispatch loop mirrors the aether coordinator verbatim:
 *   - `inFlight: (Promise<Result> | null)[]` indexed by station slot
 *   - `Promise.race(activePromises)` settles the next completion
 *   - On NaN poison we stop dispatching new work and drain in-flight ahead
 *
 * Single-station case: chunkSize = P, exactly ceil(P/B) = 1 call per chunk,
 * no inter-station hops. Generation falls back to per-token chunkSize=1.
 */

// Pipeline dispatch trace: set PIPELINE_TRACE=1 to log every dispatch/completion.
// Uses process.stderr.write when available (Node/Bun); no-op on CF Workers.
const TRACE = typeof process !== 'undefined' && process.env?.['PIPELINE_TRACE'] === '1';

// ─────────────────────────────────────────────────────────────────────────
// Amplituhedron coordinator (Death #2 — static volume cache)
// ─────────────────────────────────────────────────────────────────────────
//
// Wire layer for the host side of `AMPLITUHEDRON_PLAN.md` §4. The station
// endpoints `/amplituhedron/replay` and `/amplituhedron/capture` are live in
// `apps/distributed-inference-worker/src/index.ts:1880-1970`; this module
// fans them out across all stations from the host. Default ON; disable with
// `GNOSIS_AMPLITUHEDRON_COORDINATOR=0`.
//
// Per-station call surface is duck-typed (optional structural methods on
// IStation) so existing implementations (Station/StationUds/StationFlow/etc.)
// continue to satisfy the contract without an interface change. A station
// that does NOT implement these methods is treated as a permanent miss —
// the coordinator falls back to pipelinedPrefill as if the cache were empty.

/** Default ON. `GNOSIS_AMPLITUHEDRON_COORDINATOR=0|false` disables. */
function amplituhedronCoordinatorEnabled(): boolean {
  if (typeof process === 'undefined') return true;
  const v = process.env?.['GNOSIS_AMPLITUHEDRON_COORDINATOR'];
  if (v === undefined) return true;
  return !(v === '0' || v.toLowerCase() === 'false');
}

/**
 * Prompt-lookup / n-gram speculative decoding. DEFAULT OFF — when unset the
 * decode path is byte-identical to today. Enable with `MESH_SPEC_DECODE=1`
 * (env) or the per-Pipeline `specDecode` config override (query param → config).
 */
function meshSpecDecodeEnabledFromEnv(): boolean {
  if (typeof process === 'undefined') return false;
  const v = process.env?.['MESH_SPEC_DECODE'];
  if (v === undefined) return false;
  return v === '1' || v.toLowerCase() === 'true';
}

/** Draft length K for n-gram spec-decode. `MESH_SPEC_DECODE_K` env, default 4. */
function meshSpecDecodeKFromEnv(): number {
  if (typeof process === 'undefined') return 4;
  const v = process.env?.['MESH_SPEC_DECODE_K'];
  if (v === undefined) return 4;
  const n = Number.parseInt(v, 10);
  return Number.isInteger(n) && n >= 1 && n <= 8 ? n : 4;
}

/** Default for the spec-decode adaptive cost-gate knobs (see below). */
const SPEC_DECODE_DEFAULT_MIN_MATCH = 3;
const SPEC_DECODE_DEFAULT_COST_RATIO = 3.2;

/**
 * Minimum recurring-suffix match length before we DRAFT (a cost gate on draft
 * eligibility). `MESH_SPEC_DECODE_MIN_MATCH` env, default 3. A shorter n-gram
 * match (e.g. a single repeated token) is a weak signal: it draws a K-token
 * `/verify-batch` (~3.2× a single-token step here) that the model usually
 * rejects (accepted=0), so the round emits ONE token for K-token cost — a net
 * loss. Requiring a longer prior match means only high-continuation-probability
 * (high-acceptance) repeats draft; everything else takes the cheap 1-token step.
 * Min 1 (any match), no upper bound (clamped to the matcher's nMax in practice).
 */
function meshSpecDecodeMinMatchFromEnv(): number {
  if (typeof process === 'undefined') return SPEC_DECODE_DEFAULT_MIN_MATCH;
  const v = process.env?.['MESH_SPEC_DECODE_MIN_MATCH'];
  if (v === undefined) return SPEC_DECODE_DEFAULT_MIN_MATCH;
  const n = Number.parseInt(v, 10);
  return Number.isInteger(n) && n >= 1 ? n : SPEC_DECODE_DEFAULT_MIN_MATCH;
}

/**
 * The measured verify-cost ratio R = tK / t1 (K-candidate verify cost over a
 * single-token step). `MESH_SPEC_DECODE_COST_RATIO` env, default 3.2 (live L4
 * mistral-7b: tK≈242ms for K=4, t1≈76ms → R≈3.2). A drafted round emits
 * (accepted+1) tokens for cost tK; a 1-token round emits 1 for cost t1. Drafting
 * wins iff tK/(E[a]+1) < t1, i.e. E[a] > R - 1 (the break-even acceptance). The
 * adaptive gate suppresses drafting whenever the EWMA acceptance falls below
 * R-1. Clamped to >= 1.0 (R<1 would mean a batched verify is cheaper than a
 * single token, which is impossible — drafting would then always win).
 */
function meshSpecDecodeCostRatioFromEnv(): number {
  if (typeof process === 'undefined') return SPEC_DECODE_DEFAULT_COST_RATIO;
  const v = process.env?.['MESH_SPEC_DECODE_COST_RATIO'];
  if (v === undefined) return SPEC_DECODE_DEFAULT_COST_RATIO;
  const n = Number.parseFloat(v);
  return Number.isFinite(n) && n >= 1 ? n : SPEC_DECODE_DEFAULT_COST_RATIO;
}

/**
 * Whether the self-calibrating adaptive cost-gate is ON (default 1/ON). When
 * OFF (`MESH_SPEC_DECODE_ADAPTIVE=0`) spec-decode reverts to the prior
 * "draft on any match at fixed K" behavior (only useful for A/B measurement).
 * The gate is purely a performance decision and NEVER changes which tokens are
 * emitted — candidates=[] always degrades to one exact greedy step — so this
 * knob cannot affect correctness, only throughput.
 */
function meshSpecDecodeAdaptiveFromEnv(): boolean {
  if (typeof process === 'undefined') return true;
  const v = process.env?.['MESH_SPEC_DECODE_ADAPTIVE'];
  if (v === undefined) return true;
  return !(v === '0' || v.toLowerCase() === 'false');
}

/**
 * Compute a stable u64 prefix hash for `tokens[0..N]` as a decimal string.
 *
 * The plan (§4) names `blake3` but the host has no blake3 dep available
 * (`@a0n/aeon` doesn't expose one; `crypto-hash` isn't depended on either).
 * SHA-256 → first 8 bytes → u64 satisfies the Lean ledger's §8 "exact
 * byte-equal token sequences" requirement: distinct token streams map to
 * distinct u64s with probability ≈ 1 - 2^-64. The wire encoding is a
 * decimal string because JSON BigInt is not round-trippable.
 *
 * Cross-platform: `crypto.subtle.digest` is available in Node 22, Bun, CF
 * Workers, and modern browsers; no import required.
 */
export async function prefixHashU64(tokens: Uint32Array, domain?: string): Promise<string> {
  // Always materialize the token bytes into a fresh ArrayBuffer-backed view:
  // WebCrypto rejects SharedArrayBuffer, and domain-separation needs a
  // contiguous buffer we own.
  const tokenBytes = new Uint8Array(tokens.byteLength);
  tokenBytes.set(new Uint8Array(tokens.buffer, tokens.byteOffset, tokens.byteLength));
  // MODEL DOMAIN SEPARATION: residuals/KV are weight-specific, so a prefix
  // residual captured under one model must never be replayed for another that
  // merely shares prefix tokens. The consolidated mesh swaps models on ONE
  // service, so a model-agnostic key cross-contaminates. Prepend `<domain>\0`
  // before the token bytes → distinct models map to distinct prefix hashes.
  // Absent domain = legacy token-only hash (byte-identical to before).
  // Freshly-allocated, ArrayBuffer-backed (never SharedArrayBuffer) so WebCrypto
  // `digest` accepts it as a BufferSource.
  let bytesIn: Uint8Array<ArrayBuffer>;
  if (domain) {
    const dom = new TextEncoder().encode(domain);
    bytesIn = new Uint8Array(dom.length + 1 + tokenBytes.length);
    bytesIn.set(dom, 0);
    bytesIn[dom.length] = 0x00;
    bytesIn.set(tokenBytes, dom.length + 1);
  } else {
    bytesIn = tokenBytes;
  }
  // Sovereign bitwise XXH64 fingerprint (8 bytes / 64-bit). Faster than SHA-256
  // (no `crypto.subtle` round-trip) and pure-JS, so it works identically on every
  // host (Node/Bun/CF Workers/browser) — no FNV divergence fallback needed. Returns
  // the u64 as a decimal string, wire-compatible with the prior SHA-256-low-64 key
  // (the value changes, so the amplituhedron cache cold-starts once after deploy;
  // capture + replay share this one function so they stay consistent). Unifies the
  // transformer fleet's prefix key with the SSM fleet's bitwise key.
  const hex = cacheFingerprint64HexSync(bytesIn);
  return BigInt(`0x${hex}`).toString();
}

/**
 * Optional structural surface stations may implement to participate in the
 * Amplituhedron volume cache. Stations that omit these are treated as
 * permanent misses and the coordinator silently falls back to prefill.
 */
interface AmplituhedronReplayResult {
  hit: boolean;
  /**
   * Tail residual at position `prefixLen - 1`, shape `[hiddenDim]`. Only the
   * EXIT station is required to populate this on a hit — mid-pipeline
   * stations may return `undefined` (their KV-cache reseed is the
   * load-bearing side effect; their hit/miss flag is what the coordinator
   * gates on).
   */
  tailResidual?: Float32Array;
  tailFingerprint?: string;
  /**
   * Optional full residual stream at this station's output boundary, shape
   * `[prefixLen * hiddenDim]`. When a contiguous replay prefix provides this,
   * the host can recompute only the first missed station and its suffix.
   */
  prefixResidual?: Float32Array;
  prefixFingerprint?: string;
}

interface AmplituhedronReplayDecision {
  hit: boolean;
  tailResidual?: Float32Array;
  partialStartNodeIdx?: number;
  partialResidual?: Float32Array;
  partialFingerprint?: string;
  xorCertificate: TopologicalXorCertificate;
}

/**
 * Returns whether issue Amplituhedron Replay Xor Certificate is true.
 */
export function issueAmplituhedronReplayXorCertificate(args: {
  prefixHash: string;
  prefixLen: number;
  stationHits: readonly boolean[];
  roundSuffix?: string;
}): TopologicalXorCertificate {
  const activeLaneIndices = args.stationHits.flatMap((hit, laneIndex) =>
    hit ? [] : [laneIndex],
  );
  const mask = packedGrassmannianLaneMaskFromActiveLanes(
    args.stationHits.length,
    activeLaneIndices,
  );
  return issueGrassmannianXorRoundCertificateFromMask({
    roundId: `${args.prefixHash}:${args.prefixLen}:${args.roundSuffix ?? 'replay'}`,
    chart: 'distributed-inference-amplituhedron',
    mask,
    theoremRefs: ['Gnosis.GodBracket.god_gap_certificate'],
  });
}

interface AmplituhedronCapableStation {
  amplituhedronReplay?(args: {
    prefixHash: string;
    prefixLen: number;
    layerLo: number;
    layerHi: number;
    requestId?: string;
  }): Promise<AmplituhedronReplayResult>;
  amplituhedronCapture?(args: {
    prefixHash: string;
    prefixLen: number;
    layerLo: number;
    layerHi: number;
    tailResidual?: Float32Array;
    requestId?: string;
  }): Promise<void>;
}

function asAmplituhedronCapable(s: unknown): AmplituhedronCapableStation {
  return s as AmplituhedronCapableStation;
}

/**
 * Schedule a fire-and-forget side effect. Uses `ctx.waitUntil` when a CF
 * Worker `ExecutionContext` is reachable on `globalThis.__pipelineCtx` (the
 * host shim sets this when bootstrapping inside a Worker request);
 * otherwise falls back to `Promise.allSettled` + a stderr log on failure.
 * The promise itself is never awaited by the coordinator — the capture path
 * MUST NOT block decode.
 */
function fireAndForget(promise: Promise<unknown>, label: string): void {
  const g = globalThis as unknown as {
    __pipelineCtx?: { waitUntil(p: Promise<unknown>): void };
  };
  const ctx = g.__pipelineCtx;
  if (ctx && typeof ctx.waitUntil === 'function') {
    ctx.waitUntil(
      promise.catch((err: unknown) => {
        traceLog(`AMPLITUHEDRON ${label} fire-and-forget failed: ${String(err)}`);
      }),
    );
    return;
  }
  void Promise.allSettled([promise]).then((results) => {
    const r = results[0]!;
    if (r.status === 'rejected') {
      traceLog(`AMPLITUHEDRON ${label} fire-and-forget failed: ${String(r.reason)}`);
    }
  });
}


function traceLog(msg: string): void {
  if (!TRACE) return;
  try { process.stderr.write(`[PIPE_TRACE] ${msg}\n`); } catch { /* CF Worker — no stderr */ }
}
function traceFirst4(arr: Float32Array, label: string): string {
  if (!TRACE) return '';
  const f = [arr[0], arr[1], arr[2], arr[3]].map(v => (v ?? 0).toFixed(6));
  return ` ${label}_first4=[${f.join(',')}]`;
}

/**
 * Sovereign-sieve observability. Emitted once per armed probe on `stderr` (Cloud
 * Run captures it) so a canary shows, without a rebuild, WHICH prompt-hidden path
 * ran (native single-call `/generate` vs the host-orchestrated `computePromptHidden`
 * fallback), the station count that selected it, and whether the semantic KEY came
 * back usable (null vs a finite length + first value). Gated on the same flag that
 * arms the sieve (`MOONSHINE_SOVEREIGN_SIEVE` = 1/true/on) so it is zero-cost off.
 * This is the instrument that turns "the lane is silently inert" into a one-line
 * diagnosis (path + stations + hidden-null).
 */
function logSievePromptHiddenPath(
  path: 'native' | 'fallback',
  stations: number,
  promptHidden: Float32Array | null,
): void {
  const raw = typeof process !== 'undefined' ? process.env?.['MOONSHINE_SOVEREIGN_SIEVE'] : undefined;
  if (raw === undefined) return;
  const s = raw.trim().toLowerCase();
  if (s !== '1' && s !== 'true' && s !== 'on') return;
  const hidden = promptHidden ? `len=${promptHidden.length} first=${(promptHidden[0] ?? 0).toFixed(6)}` : 'NULL';
  try {
    process.stderr.write(`[SIEVE] promptHidden path=${path} stations=${stations} hidden=${hidden}\n`);
  } catch { /* CF Worker — no stderr */ }
}

import { getTeleportationOptimizer, type TeleportationOptimizer } from '@a0n/aeon/optimization';
import { cacheFingerprint64HexSync } from '@a0n/bitwise/cache-fp64';
import {
  applyTerminalProsodyBias,
  argmaxTerminalCandidate,
  emitTerminalSelectionOutcome,
  reynoldsPhiBreak,
  type TerminalProsodyContext,
  type TerminalProsodyEvent,
  type TerminalProsodyMode,
  type TerminalTailState,
} from '@a0n/aether/terminal-prosody';
import {
  AlphaJumpDispatcher,
  FANO_LINES,
  FanoLine,
  type ResidualSeedAck,
} from './alpha-jump';
import {
  requestEntropy,
  type SpectralNoiseMonitor,
} from './spectral-noise-monitor';
import { resolveMahaShardRoute } from './amplituhedron-sharding';
import { ngramDraft, ngramDraftWithMatch } from './ngram-draft';
import type { IStation, PersonalitySteeringSpec } from './types';
import {
  issueGrassmannianXorRoundCertificateFromMask,
  packedGrassmannianLaneMaskFromActiveLanes,
  type TopologicalXorCertificate,
} from '../../src/topological-xor-certificate';

// ─────────────────────────────────────────────────────────────────────────
// Death #4 — Octonion Fano-plane routing (canonical-chain wire-in)
// ─────────────────────────────────────────────────────────────────────────
//
// `FANO_FULL_MASK` — the receiver's "fully resolved" sentinel. Mirrors
// `FANO_FULL_MASK` in `apps/worker-node1/src/residual-seed-handler.ts:109`
// (also `apps/distributed-inference-worker` and `apps/worker-node2`). Lives
// here as a const because alpha-jump.ts does NOT export it (the dispatcher
// is line-symmetric and never inspects the bitset itself — only the gate
// caller does, which is exactly this file).
//
// 0x7f === bits 0..6 set === every line in `FANO_LINES` ticked.
const FANO_FULL_MASK = (1 << FANO_LINES.length) - 1;

/**
 * Optional opt-in hook that lets a coordinator consult Death #4's
 * `AlphaJumpDispatcher` after each per-token canonical hop. Default OFF —
 * existing callers (and production) must explicitly construct an
 * `AlphaJumpHook` to participate. This is the safety net behind the
 * "the floor is no worse than today" guarantee from
 * `OCTONION_ROUTING_PLAN.md` §5.
 *
 * CAVEAT — Fano-sum reconstruction (gap G3, NOT yet closed):
 * --------------------------------------------------------------
 * The receiver's `/residual-seed` ack (see
 * `apps/worker-node1/src/residual-seed-handler.ts:396-408`) returns
 * `{ shadow_mode, fully_resolved, awaiting_resolution?, bitset, ... }`
 * but does NOT carry the accumulated f32 buffer back to the coordinator.
 * The Fano sum lives entirely on the receiver's per-token slot. Two
 * possible wire futures:
 *   (a) extend the ack to embed the slot's f32 sum (large body, but
 *       lets the coordinator skip the next station.forward() call and
 *       feed the sum into station N+2 directly), OR
 *   (b) arrange for the receiver's NEXT internal compute (`/split-a`
 *       or its `/forward` peer) to consume its slot when it fires —
 *       in which case the coordinator's "skip" is fully implicit and
 *       the ack only needs to gate the next dispatch.
 *
 * Until one of those lands, `fanoSubstitute` is the operator's escape
 * hatch: when the gate fires AND a substitute is provided, the
 * coordinator skips the next canonical hop and uses the substitute as
 * the residual fed into station N+2. When undefined, the coordinator
 * STILL never skips (defensive default) — it only logs the would-skip
 * via the dispatcher's CONSPICUOUS LOG line. Tests use the substitute
 * to exercise the skip branch without requiring the wire-format
 * extension to ship first.
 *
 * Bench-fano-parity (gap G2/G3, see OCTONION_ROUTING_PLAN.md §1) MUST
 * confirm zero divergence between Fano sum and canonical residual
 * before `fanoSubstitute` is wired in production. This file does not
 * police that — the operator who supplies the substitute does.
 */
export interface AlphaJumpHook {
  dispatcher: AlphaJumpDispatcher;
  /**
   * Maps station index N → destination URL for `/residual-seed` POSTs.
   * The dispatcher fires from station N to station N+1's port. Returning
   * `undefined` for any nodeIdx makes the coordinator skip dispatch on
   * that hop (treats the station as not-Fano-routable).
   *
   * prefixHash is supplied to enable Maha-anchored sharding (mod 21)
   * in multi-node AlphaJump clusters.
   */
  resolveStationUrl(nodeIdx: number, prefixHash?: string): string | undefined;
  /**
   * Maps Fano line + src layer → dst layer. Defaults to Pair X
   * (`AlphaJumpDispatcher.computeJumpDestination(FanoLine.E1_E4_E5, src)`)
   * — the skip-attention shortcut, which is the only line worth a
   * standalone implementation per OCTONION_ROUTING_PLAN.md §3.
   */
  computeDispatch?(nodeIdx: number, srcLayer: number): {
    line: FanoLine;
    sign: 'pos' | 'neg';
    dstLayer: number;
  } | null;
  /**
   * When the gate authorizes skipping the canonical hop, callers MAY
   * supply the substitute residual (see CAVEAT block above). Signature:
   * `(ack, currentResidual) => substituteResidual | undefined`. Returning
   * `undefined` aborts the skip — the coordinator falls back to the
   * canonical chain. Default behavior (no substitute supplied): never
   * skip, regardless of gate.
   */
  fanoSubstitute?(
    ack: ResidualSeedAck,
    currentResidual: Float32Array,
  ): Float32Array | undefined;
  /**
   * Optional per-dispatch fetch override for `dispatchSeedHttp`. Falls
   * through to `globalThis.fetch` when omitted.
   */
  fetchImpl?: typeof fetch;
  /**
   * Optional `RequestInit` extras forwarded to the dispatcher (e.g.
   * Authorization headers, abort signals). The dispatcher merges
   * `headers` and ignores `body`/`method`.
   */
  requestInit?: Partial<RequestInit>;
}

export interface PipelineSamplingContext {
  emittedTokens: number;
  maxNewTokens: number;
  generatedTokens: readonly number[];
  tailState: TerminalTailState;
  stopTokens: ReadonlySet<number>;
}

export interface PipelinePrefillProgress {
  completedTokens: number;
  totalTokens: number;
  completedChunks: number;
  totalChunks: number;
}

export interface PipelineTokenEvent {
  tokenId: number;
  tokenIndex: number;
  maxNewTokens: number;
}

export type PipelineTerminalProsodyConfig =
  Omit<
    Partial<TerminalProsodyContext>,
    'emittedTokens' | 'maxTokens' | 'tailState' | 'stopTokenIds'
  > & {
    mode?: TerminalProsodyMode;
    tailState?: TerminalTailState;
    stopTokenIds?: Iterable<number>;
  };

export interface PipelineConfig {
  /** Stations ordered by layer range; first is entry, last is exit. */
  stations: IStation[];
  /** Optional explicit chunkSize override. Default: max(1, floor(P / N)). */
  chunkSize?: number;
  /** Hidden dim of the model. Required so we can slice contiguous chunks. */
  hiddenDim: number;
  /** Vocab size of the model. Used to size logits returned by lmHead. */
  vocabSize?: number;
  /**
   * Optional sampler. If unspecified, argmax sampling is used. Receives the
   * logits Float32Array for the most recent token and returns the chosen
   * token id.
   */
  sample?: (logits: Float32Array, context: PipelineSamplingContext) => number;
  /** Optional finite-horizon terminal pressure controller. */
  terminalProsody?: PipelineTerminalProsodyConfig;
  /** Optional request id seed for tracing across station calls. */
  requestId?: string;
  /**
   * Model identity used to domain-separate the Amplituhedron volume-cache key.
   * Residuals are weight-specific; scoping the prefix hash by model prevents a
   * residual captured under one model from being replayed for another on a
   * shared/long-lived station (the consolidated mesh swaps models on one
   * service). Absent = legacy model-agnostic key.
   */
  amplituhedronModelKey?: string;
  /** Enable Maha-anchored sharding for Amplituhedron volumes (default: false). */
  amplituhedronSharding?: boolean;
  /**
   * Optional passive spectral-noise monitor. When configured, the pipeline
   * emits log-only layer-node timing traces; no routing or sharding decision
   * is changed by the assessment.
   */
  spectralNoiseMonitor?: SpectralNoiseMonitor;
  /**
   * Enable the bizarro companion-mask speculative path. When true, idle
   * downstream pipeline slots are used to fire fire-and-forget reject-mask
   * forwards on station 0 against the next-unstarted chunk. If a mask is
   * ready by the time the chunk retires AND `argmax(nominal_logits)` is
   * outside the mask, the chunk retires "early" (counted in
   * `bizarroRetirements`; nominal logits are still authoritative).
   *
   * Mirrors `enableBizarroPrefill` from
   * `open-source/aether/src/gcp-coordinator.ts`. Backed by
   * `BizarroCheckpointHomology.lean` and `AeonFlowBizarro.lean`.
   *
   * Default false. Speculative errors never propagate — bizarro work is
   * fire-and-forget and cannot poison the nominal path.
   */
  enableBizarro?: boolean;
  /**
   * TQFT Cobordism: vocabulary-sharded embed stations.
   * When present, pipelinedPrefill replaces the single entryStation.embed()
   * with a parallel fan-out to all cobordism shards, merging partial
   * embeddings via elementwise addition (Frobenius μ: A ⊗ A → A).
   *
   * Each entry maps a Station to its vocabulary range.
   */
  cobordismStations?: Array<{
    station: IStation;
    vocabStart: number;
    vocabEnd: number;
  }>;
  /**
   * Statistical Teleportation: skip computation for converged ("dead") layers.
   * When enabled, the pipeline uses per-layer deficits from station responses
   * to determine which layer ranges can be safely bypassed.
   *
   * Default false.
   */
  enableTeleportation?: boolean;
  /**
   * Per-chunk fano cache (HOT PATH). `station.forward(residual,startPos,seqLen)`
   * is a pure function, so content-addressing it lets a repeated prefix replay at
   * `geodesicLength:0` — skipping the (often bound-service) forward entirely. Hits
   * are FREE (zero token burn); forwards burn `chunk.count` tokens for recoup.
   * Default ON (additive memo on a pure fn; set `false` to disable).
   */
  enableChunkCache?: boolean;
  /**
   * Disable the single-station native /generate shortcut. Some GGUF
   * quantization kernels can serve host-orchestrated /embed + /forward decode
   * even when their native batched generator is not usable.
   */
  disableNativeGenerate?: boolean;
  /**
   * Optional stop tokens. When the sampler picks one of these IDs,
   * `Pipeline.generate()` returns early (before reaching `maxNewTokens`).
   */
  stopTokens?: number[];
  /**
   * Prompt-lookup / n-gram speculative decoding (greedy, lossless). DEFAULT OFF.
   * When undefined, falls back to the `MESH_SPEC_DECODE` env flag (also default
   * OFF), so unset config + unset env = byte-identical to the non-speculative
   * decode path. Per-request callers (e.g. a `?spec=1` query param) can set this
   * explicitly to override the env. Greedy only: the loop is skipped when a
   * sampler / personality steering / terminal-prosody / multi-station path is
   * active (those are not greedy-argmax and so are not lossless under this rule).
   */
  specDecode?: boolean;
  /** Draft length K for n-gram spec-decode (1..8). Default: env or 4. */
  specDecodeK?: number;
  /**
   * Self-calibrating adaptive cost-gate for spec-decode (default: env or ON).
   * When ON, the loop only drafts when the n-gram match is strong enough AND the
   * rolling acceptance clears the verify-cost break-even, and it shrinks/grows K
   * with acceptance — guaranteeing spec-decode is never materially slower than
   * the baseline on low-acceptance (prose/counting) workloads while keeping the
   * repetitive-text win. Purely a performance gate: it never changes the emitted
   * tokens (lossless). When OFF, drafts on any match at fixed K (legacy).
   */
  specDecodeAdaptive?: boolean;
  /**
   * Minimum recurring-suffix match length before drafting (default: env or 3).
   * See `MESH_SPEC_DECODE_MIN_MATCH`. Only used when the adaptive gate is ON.
   */
  specDecodeMinMatch?: number;
  /**
   * Measured verify-cost ratio R = tK/t1 (default: env or 3.2). Drafting wins
   * only when EWMA acceptance > R - 1. See `MESH_SPEC_DECODE_COST_RATIO`. Only
   * used when the adaptive gate is ON.
   */
  specDecodeCostRatio?: number;
}

export interface PrefillResult {
  /** Final-residual / lm-head logits for the last prompt token. */
  logits: Float32Array;
  /** ceil(P/B) + (N-1) -- the number of station calls the pipeline issued. */
  stationCallCount: number;
  /** True if the pipeline aborted early on NaN poison. */
  poisoned: boolean;
  /** Chunk index where poison was first observed; -1 otherwise. */
  poisonChunk: number;
  /** Tail residual used for lm_head, shape `[hiddenDim]`. */
  tailResidual?: Float32Array;
  /** Per-station full prefix residual streams, each shape `[prefixLen * hiddenDim]`. */
  stationPrefixResiduals?: readonly Float32Array[];
  /**
   * Number of speculative bizarro companions fired during prefill. Only
   * present when `enableBizarro: true`. Always 0 or absent on the nominal
   * path.
   */
  bizarroDispatches?: number;
  /**
   * Number of chunks that satisfied the bizarro early-retirement rule
   * (mask ready AND `argmax(nominal_logits) ∉ mask`). Counter only — the
   * nominal logits remain authoritative; this is observability for the
   * "skip recheck" semantics that multi-token verification will eventually
   * cash in. Only present when `enableBizarro: true`.
   */
  bizarroRetirements?: number;
}

export type PipelinePrefillSource =
  | 'amplituhedron-hit'
  | 'amplituhedron-partial'
  | 'amplituhedron-miss'
  | 'disabled';

export interface PipelinePrefillTelemetry {
  source: PipelinePrefillSource;
  prefixHash?: string;
  prefixLen: number;
  replayMs: number;
  prefillMs: number;
  captureScheduled: boolean;
  partialStartNodeIdx?: number;
  xorStatus?: TopologicalXorCertificate['status'];
  xorDefectRank?: number;
  stationCallCount: number;
}

export interface PipelinePreparePrefillArgs {
  promptTokens: Uint32Array;
  onPrefillProgress?: (progress: PipelinePrefillProgress) => void;
  /**
   * Skip the lm_head / logits forward when false (e.g. maxNewTokens === 0,
   * admission/cache-warm probes). Defaults to true. Mirrors the
   * `computeLogits` knob on `pipelinedPrefill`/`prefillFromReplaySuffix` so
   * the prefill+capture still warms the Amplituhedron cache without paying
   * for a logits forward that nothing will consume.
   */
  computeLogits?: boolean;
}

export interface PipelinePreparedPrefill {
  prefill: PrefillResult;
  telemetry: PipelinePrefillTelemetry;
}

export interface PipelineGenerateFromPrefillArgs {
  promptTokens: Uint32Array;
  prefixHash?: string;
  prefill: PrefillResult;
  maxNewTokens: number;
  onToken?: (event: PipelineTokenEvent) => void;
  /**
   * Death #4 opt-in. See `Pipeline.generate` for the contract; mirrored
   * here so callers that drive prefill themselves can still feed the
   * dispatcher into the per-token decode loop.
   */
  alphaJump?: AlphaJumpHook;
}

interface ChunkDescriptor {
  startPos: number;
  count: number;
}

interface InFlightResult {
  nodeIdx: number;
  chunkIdx: number;
  result: Float32Array;
}

interface AmplituhedronStationWindow {
  start: number;
  end: number;
  sharded: boolean;
}

export class Pipeline {
  private readonly stations: IStation[];
  private readonly hiddenDim: number;
  private readonly explicitChunkSize: number | undefined;
  private readonly sample: (logits: Float32Array, context: PipelineSamplingContext) => number;
  private readonly usesDefaultSampler: boolean;
  private readonly requestId: string;
  private readonly amplituhedronModelKey?: string;
  private readonly enableBizarro: boolean;
  private readonly cobordismStations:
    | Array<{ station: IStation; vocabStart: number; vocabEnd: number }>
    | undefined;
  private readonly enableTeleportation: boolean;
  // Per-chunk fano cache (hot path): pure-forward memo for geodesicLength:0 replay.
  private readonly enableChunkCache: boolean;
  private readonly disableNativeGenerate: boolean;
  /** Set while a per-request personality steering filter is installed. */
  private steeringActive = false;
  private readonly chunkCacheCap = 8192;
  private readonly chunkCache = new Map<string, Float32Array>();
  private chunkCacheHits = 0;
  private chunkCacheMisses = 0;
  private chunkTokenBurn = 0;
  // Amplituhedron volume-cache hit/miss tallies (process lifetime). Exposed via
  // getCacheStats() + GET /cache/stats so COST_MODEL.md's blended cost can be MEASURED
  // (cache-hit-rate is the master variable) instead of assumed. Counts only; no payloads.
  private amplituhedronHits = 0;
  private amplituhedronPartials = 0;
  private amplituhedronMisses = 0;
  private readonly teleportationOptimizer: TeleportationOptimizer | undefined;
  private readonly amplituhedronSharding: boolean;
  private readonly spectralNoiseMonitor: SpectralNoiseMonitor | undefined;
  private readonly stopTokens: Set<number>;
  private readonly terminalProsody: PipelineTerminalProsodyConfig | undefined;
  private bypassedLayers: Set<number> = new Set();
  /** Prompt-lookup / n-gram speculative decode (greedy, lossless). DEFAULT OFF. */
  private readonly specDecode: boolean;
  private readonly specDecodeK: number;
  /** Adaptive cost-gate ON/OFF + its two knobs (min match length, cost ratio). */
  private readonly specDecodeAdaptive: boolean;
  private readonly specDecodeMinMatch: number;
  private readonly specDecodeCostRatio: number;

  constructor(config: PipelineConfig) {
    if (!config.stations.length) {
      throw new Error('Pipeline: at least one station is required');
    }
    if (!Number.isInteger(config.hiddenDim) || config.hiddenDim <= 0) {
      throw new Error(
        `Pipeline: hiddenDim must be a positive integer (got ${String(config.hiddenDim)})`,
      );
    }
    if (
      config.chunkSize !== undefined &&
      (!Number.isInteger(config.chunkSize) || config.chunkSize <= 0)
    ) {
      throw new Error(
        `Pipeline: chunkSize must be a positive integer if set (got ${String(config.chunkSize)})`,
      );
    }
    this.stations = config.stations.slice();
    this.hiddenDim = config.hiddenDim;
    this.explicitChunkSize = config.chunkSize;
    this.usesDefaultSampler = config.sample === undefined;
    this.sample = config.sample ?? argmax;
    this.requestId = config.requestId ?? `pipe-${Date.now()}`;
    this.amplituhedronModelKey = config.amplituhedronModelKey;
    this.enableBizarro = config.enableBizarro === true;
    this.amplituhedronSharding = config.amplituhedronSharding === true;
    this.cobordismStations = config.cobordismStations?.slice();
    this.enableTeleportation = config.enableTeleportation === true;
    this.enableChunkCache = config.enableChunkCache !== false; // default ON (hot-path teleport)
    this.disableNativeGenerate = config.disableNativeGenerate === true;
    this.spectralNoiseMonitor = config.spectralNoiseMonitor;
    this.stopTokens = new Set(config.stopTokens ?? []);
    this.terminalProsody = config.terminalProsody;
    // Spec-decode: explicit config wins; otherwise env flag (default OFF).
    this.specDecode = config.specDecode ?? meshSpecDecodeEnabledFromEnv();
    this.specDecodeK =
      config.specDecodeK !== undefined &&
      Number.isInteger(config.specDecodeK) &&
      config.specDecodeK >= 1 &&
      config.specDecodeK <= 8
        ? config.specDecodeK
        : meshSpecDecodeKFromEnv();
    this.specDecodeAdaptive =
      config.specDecodeAdaptive ?? meshSpecDecodeAdaptiveFromEnv();
    this.specDecodeMinMatch =
      config.specDecodeMinMatch !== undefined &&
      Number.isInteger(config.specDecodeMinMatch) &&
      config.specDecodeMinMatch >= 1
        ? config.specDecodeMinMatch
        : meshSpecDecodeMinMatchFromEnv();
    this.specDecodeCostRatio =
      config.specDecodeCostRatio !== undefined &&
      Number.isFinite(config.specDecodeCostRatio) &&
      config.specDecodeCostRatio >= 1
        ? config.specDecodeCostRatio
        : meshSpecDecodeCostRatioFromEnv();
    if (this.enableTeleportation) {
      this.teleportationOptimizer = getTeleportationOptimizer();
    }
  }

  private canShardAmplituhedronStations(): boolean {
    if (this.stations.length <= 1) return true;
    const first = this.stations[0]!;
    return this.stations.every(
      (station) =>
        station.startLayer === first.startLayer &&
        station.endLayer === first.endLayer,
    );
  }

  private resolveAmplituhedronStationWindow(prefixHash: string): AmplituhedronStationWindow {
    if (!this.amplituhedronSharding || !this.canShardAmplituhedronStations()) {
      return { start: 0, end: this.stations.length, sharded: false };
    }
    const route = resolveMahaShardRoute(this.stations.length, prefixHash);
    if (!route) return { start: 0, end: this.stations.length, sharded: false };
    return { start: route.start, end: route.end, sharded: true };
  }

  /** First (entry) station -- runs token embedding. */
  get entryStation(): IStation {
    return this.stations[0]!;
  }

  /** Last (exit) station -- runs final norm + lm_head. */
  get exitStation(): IStation {
    return this.stations[this.stations.length - 1]!;
  }

  /**
   * Token-embedding lookup ONLY (no layer forward, no generation): map a chunk
   * of tokens to the initial residual stream of shape `[numTokens * hiddenDim]`.
   *
   * When cobordism (vocab-sharded `token_embd_weight`) stations are configured,
   * fans out `/cobordism-embed` to every shard in parallel and Frobenius-merges
   * (elementwise add of disjoint per-token partials). Otherwise falls back to a
   * single entry-station `/embed`. This is fast — one embedding-table lookup,
   * no attention/MLP matmuls and no autoregressive decode — so it is the right
   * primitive for sentence embeddings.
   */
  async embed(promptTokens: Uint32Array, startPos = 0): Promise<{ embeddings: Float32Array; hiddenDim: number; numTokens: number }> {
    const numTokens = promptTokens.length;
    const hiddenDim = this.hiddenDim;
    const totalFloats = numTokens * hiddenDim;
    let embeddings: Float32Array;
    if (this.cobordismStations && this.cobordismStations.length > 0) {
      const partials = await Promise.all(
        this.cobordismStations.map((cs) =>
          cs.station.cobordismEmbed({
            tokens: promptTokens,
            startPos,
            vocabStart: cs.vocabStart,
            vocabEnd: cs.vocabEnd,
            requestId: this.requestId,
          }),
        ),
      );
      embeddings = new Float32Array(totalFloats);
      for (const partial of partials) {
        if (partial.length !== totalFloats) {
          throw new Error(
            `Pipeline.embed cobordism merge: shard returned ${partial.length} floats, expected ${totalFloats}`,
          );
        }
        for (let j = 0; j < totalFloats; j++) embeddings[j]! += partial[j]!;
      }
    } else {
      embeddings = await this.entryStation.embed({ tokens: promptTokens, startPos, requestId: this.requestId });
      if (embeddings.length !== totalFloats) {
        throw new Error(
          `Pipeline.embed: entry station returned ${embeddings.length} floats, expected ${totalFloats}`,
        );
      }
    }
    return { embeddings, hiddenDim, numTokens };
  }

  async saveKvSession(sessionId: string, prefixLength: number): Promise<boolean> {
    const savers = this.stations
      .map((station) => station.saveKvSession)
      .filter((save): save is NonNullable<IStation['saveKvSession']> => save !== undefined);
    if (savers.length !== this.stations.length) return false;

    const results = await Promise.all(
      this.stations.map((station) =>
        station.saveKvSession!({
          sessionId,
          prefixLength,
          requestId: this.requestId,
        }),
      ),
    );
    return results.every(Boolean);
  }

  async loadKvSession(sessionId: string): Promise<{ prefixLength: number } | null> {
    const loaders = this.stations
      .map((station) => station.loadKvSession)
      .filter((load): load is NonNullable<IStation['loadKvSession']> => load !== undefined);
    if (loaders.length !== this.stations.length) return null;

    const results = await Promise.all(
      this.stations.map((station) =>
        station.loadKvSession!({
          sessionId,
          requestId: this.requestId,
        }),
      ),
    );
    if (results.some((result) => result === null)) return null;
    const prefixLength = results[0]!.prefixLength;
    if (results.some((result) => result!.prefixLength !== prefixLength)) return null;
    return { prefixLength };
  }

  async deleteKvSession(sessionId: string): Promise<boolean> {
    const deleters = this.stations
      .map((station) => station.deleteKvSession)
      .filter((del): del is NonNullable<IStation['deleteKvSession']> => del !== undefined);
    if (deleters.length !== this.stations.length) return false;

    const results = await Promise.all(
      this.stations.map((station) =>
        station.deleteKvSession!({
          sessionId,
          requestId: this.requestId,
        }),
      ),
    );
    return results.every(Boolean);
  }

  private createTerminalProsodyContext(
    context: PipelineSamplingContext,
  ): TerminalProsodyContext | undefined {
    const config = this.terminalProsody;
    if (!config || config.enabled === false) return undefined;

    const stopTokenIds = new Set<number>(this.stopTokens);
    if (config.stopTokenIds) {
      for (const tokenId of config.stopTokenIds) stopTokenIds.add(tokenId);
    }

    return {
      ...config,
      emittedTokens: context.emittedTokens,
      maxTokens: context.maxNewTokens,
      tailState: context.tailState,
      stopTokenIds,
      trace: {
        ...config.trace,
        requestId: config.trace?.requestId ?? this.requestId,
        source: config.trace?.source ?? 'gnosis:distributed-inference-host',
        step: config.trace?.step ?? context.emittedTokens - 1,
      },
      onEvent: (event) => {
        config.onEvent?.(event);
        this.recordTerminalProsodyEvent(event);
      },
    };
  }

  private terminalProsodyBiasActive(): boolean {
    return (
      this.terminalProsody !== undefined &&
      this.terminalProsody.enabled !== false &&
      this.terminalProsody.mode !== 'observe'
    );
  }

  private emitObservedTerminalProsody(
    context: PipelineSamplingContext,
    selectedTokenId: number,
    sampler: string,
  ): void {
    if (this.terminalProsody?.mode !== 'observe') return;
    const terminalContext = this.createTerminalProsodyContext(context);
    if (!terminalContext || !reynoldsPhiBreak(terminalContext)) return;
    // Observe mode records the selection-outcome half of
    // ForkRaceFoldTheorems.TerminalProsody.OpenThenStopSelectionCertificate
    // without disabling the native /decode-next or /sample-next fast path.
    emitTerminalSelectionOutcome({
      context: terminalContext,
      baseline: null,
      adjusted: null,
      selectedTokenId,
      sampler,
    });
  }

  private recordTerminalProsodyEvent(event: TerminalProsodyEvent): void {
    traceLog(
      `TERMINAL_PROSODY kind=${event.kind}` +
        ` emitted=${event.emittedTokens}/${event.maxTokens}` +
        ` remaining=${event.remainingTokens}` +
        ` directive=${event.directive}` +
        ` adjusted=${event.adjustedTokenCount ?? 0}` +
        ` deferred=${event.deferredStopCount ?? 0}` +
        ` injected=${event.injectedTokenIds?.join(',') ?? ''}` +
        ` baseline=${event.baselineTokenId ?? ''}` +
        ` adjustedToken=${event.adjustedTokenId ?? ''}` +
        ` selected=${event.selectedTokenId ?? ''}` +
        ` changed=${event.argmaxChanged ?? ''}` +
        ` sampler=${event.sampler ?? ''}`,
    );
  }

  private chooseToken(
    logits: Float32Array,
    context: PipelineSamplingContext,
  ): number {
    const terminalContext = this.createTerminalProsodyContext(context);
    if (!terminalContext || !reynoldsPhiBreak(terminalContext)) {
      return this.sample(logits, context);
    }
    if (!this.terminalProsodyBiasActive()) {
      const selectedTokenId = this.sample(logits, context);
      // The formal certificate is telemetry-facing: observe reports the same
      // post-break selection-outcome pressure fields while preserving logits.
      emitTerminalSelectionOutcome({
        context: terminalContext,
        baseline: argmaxTerminalCandidate(logits),
        adjusted: argmaxTerminalCandidate(logits),
        selectedTokenId,
        sampler: 'gnosis:pipeline:observe',
      });
      return selectedTokenId;
    }
    const adjustedLogits = terminalContext ? new Float32Array(logits) : logits;
    const baseline = terminalContext
      ? argmaxTerminalCandidate(adjustedLogits)
      : null;
    if (terminalContext) {
      applyTerminalProsodyBias(adjustedLogits, terminalContext);
    }
    const adjusted = terminalContext
      ? argmaxTerminalCandidate(adjustedLogits)
      : null;
    const selectedTokenId = this.sample(adjustedLogits, context);
    if (terminalContext) {
      // Bias mode is the runtime side of OpenThenStopSelectionCertificate:
      // open tails close first; closed tails may then select a stop-maximal token.
      emitTerminalSelectionOutcome({
        context: terminalContext,
        baseline,
        adjusted,
        selectedTokenId,
        sampler: 'gnosis:pipeline',
      });
    }
    return selectedTokenId;
  }

  private canUseRemoteGreedySample(context: PipelineSamplingContext): boolean {
    if (!this.usesDefaultSampler) return false;
    if (this.cobordismStations && this.cobordismStations.length > 0) return false;
    if (!this.exitStation.sampleNext) return false;
    if (!this.terminalProsodyBiasActive()) return true;
    const terminalContext = this.createTerminalProsodyContext(context);
    return !terminalContext || !reynoldsPhiBreak(terminalContext);
  }

  private canUseRemoteDecode(context: PipelineSamplingContext): boolean {
    if (this.stations.length !== 1) return false;
    if (this.enableTeleportation) return false;
    if (!this.entryStation.decodeNext) return false;
    if (!this.usesDefaultSampler) return false;
    if (this.cobordismStations && this.cobordismStations.length > 0) return false;
    // Terminal-prosody no longer forces the slow `chooseTokenThroughStations`
    // fallback: `/decode-next` applies the SAME native cadence selection
    // (see remoteDecodeProsody + the fat-station handler), so the terminal
    // window stays on the one-call-per-token fast path. We still skip the
    // remote path in 'observe' mode (terminalProsodyBiasActive() is false but
    // we want the host-side emitObservedTerminalProsody to fire) — observe is
    // handled below by emitting from the decode-next branch.
    return true;
  }

  /**
   * Per-step native terminal-prosody params for `/decode-next`, or undefined
   * when cadence biasing is inactive (so the native side stays pure argmax).
   * `undefined` outside the terminal window too — `select_terminal_token`
   * short-circuits to baseline there, but skipping the headers keeps the wire
   * minimal and the native fast-argmax path untouched until pressure crosses.
   */
  private remoteDecodeProsody(context: PipelineSamplingContext):
    | {
        emittedTokens: number;
        maxTokens: number;
        tailState: 'open-clause' | 'sentence-closed';
        sentenceCloseTokenIds?: readonly number[];
        stopTokenIds?: readonly number[];
      }
    | undefined {
    if (!this.terminalProsodyBiasActive()) return undefined;
    const terminalContext = this.createTerminalProsodyContext(context);
    if (!terminalContext || !reynoldsPhiBreak(terminalContext)) return undefined;
    const native = this.nativeTerminalProsodyConfig();
    const result: {
      emittedTokens: number;
      maxTokens: number;
      tailState: 'open-clause' | 'sentence-closed';
      sentenceCloseTokenIds?: readonly number[];
      stopTokenIds?: readonly number[];
    } = {
      emittedTokens: context.emittedTokens,
      maxTokens: context.maxNewTokens,
      tailState: context.tailState === 'sentence-closed'
        ? 'sentence-closed'
        : 'open-clause',
    };
    if (native?.sentenceCloseTokenIds) {
      result.sentenceCloseTokenIds = native.sentenceCloseTokenIds;
    }
    if (native?.stopTokenIds) {
      result.stopTokenIds = native.stopTokenIds;
    }
    return result;
  }

  private sampleNextUnavailable(error: unknown): boolean {
    const message = error instanceof Error ? error.message : String(error);
    return (
      message.includes('/sample-next') &&
      (message.includes('HTTP 404') || message.includes('not available'))
    );
  }

  private decodeNextUnavailable(error: unknown): boolean {
    const message = error instanceof Error ? error.message : String(error);
    return (
      message.includes('/decode-next') &&
      (message.includes('HTTP 404') || message.includes('not available'))
    );
  }

  private nativeGenerateUnavailable(error: unknown): boolean {
    const message = error instanceof Error ? error.message : String(error);
    return (
      message.includes('/generate') &&
      (message.includes('HTTP 404') ||
        message.includes('not found') ||
        message.includes('not available'))
    );
  }

  private async chooseTokenFromResidual(
    residual: Float32Array,
    samplingContext: PipelineSamplingContext,
  ): Promise<number> {
    // lm_head: same cobordism comultiplication as pipelinedPrefill -- the
    // exit station does not hold output_weight, so partial logits come
    // from the cobordism shards (tied embeddings -> same row, just used as
    // a column instead of a lookup).
    let logits: Float32Array;
    if (this.cobordismStations && this.cobordismStations.length > 0) {
      const partials = await Promise.all(
        this.cobordismStations.map((cs) =>
          cs.station.cobordismLmHead({
            residual,
            sequenceLength: 1,
            vocabStart: cs.vocabStart,
            vocabEnd: cs.vocabEnd,
            requestId: this.requestId,
          }),
        ),
      );
      const totalVocab = this.cobordismStations.reduce(
        (acc, cs) => Math.max(acc, cs.vocabEnd),
        0,
      );
      logits = new Float32Array(totalVocab);
      for (let i = 0; i < this.cobordismStations.length; i++) {
        const cs = this.cobordismStations[i]!;
        const partial = partials[i]!;
        if (partial.length !== cs.vocabEnd - cs.vocabStart) {
          throw new Error(
            `Pipeline.generate cobordism lm_head: shard ${i} returned ${partial.length} floats, ` +
              `expected ${cs.vocabEnd - cs.vocabStart}`,
          );
        }
        logits.set(partial, cs.vocabStart);
      }
    } else {
      logits = await this.exitStation.lmHead({
        residual,
        sequenceLength: 1,
        requestId: this.requestId,
      });
    }
    return this.chooseToken(logits, samplingContext);
  }

  private async chooseTokenThroughStations(
    tokenId: number,
    position: number,
    samplingContext: PipelineSamplingContext,
    alphaJump?: AlphaJumpHook,
    prefixHash?: string,
  ): Promise<number | null> {
    const tokens = new Uint32Array([tokenId]);
    // Embed: cobordism fan-out + Frobenius merge when shards configured;
    // legacy entryStation.embed otherwise. (Mirrors pipelinedPrefill.)
    let initial: Float32Array;
    if (this.cobordismStations && this.cobordismStations.length > 0) {
      const partials = await Promise.all(
        this.cobordismStations.map((cs) =>
          cs.station.cobordismEmbed({
            tokens,
            startPos: position,
            vocabStart: cs.vocabStart,
            vocabEnd: cs.vocabEnd,
            requestId: this.requestId,
          }),
        ),
      );
      initial = new Float32Array(this.hiddenDim);
      for (const partial of partials) {
        if (partial.length !== this.hiddenDim) {
          throw new Error(
            `Pipeline.generate cobordism embed: shard returned ${partial.length} floats, expected ${this.hiddenDim}`,
          );
        }
        for (let j = 0; j < this.hiddenDim; j++) {
          initial[j]! += partial[j]!;
        }
      }
    } else {
      initial = await this.entryStation.embed({
        tokens,
        startPos: position,
        requestId: this.requestId,
      });
    }

    let residual = initial;
    for (let nodeIdx = 0; nodeIdx < this.stations.length; nodeIdx++) {
      const station = this.stations[nodeIdx]!;

      // Teleportation skip: if all layers in this station's range are bypassed
      if (this.enableTeleportation && this.teleportationOptimizer) {
        const bypassed = this.getBypassedStations();
        if (bypassed.has(nodeIdx)) {
          traceLog(`TELEPORT SKIP node=${nodeIdx} range=${station.startLayer}-${station.endLayer}`);
          continue;
        }
      }

      const reply = await station.forward({
        residual,
        startPos: position,
        sequenceLength: 1,
        requestId: this.requestId,
      });
      residual = reply.residual;

      if (this.enableTeleportation && this.teleportationOptimizer && reply.teleportation) {
        this.updateTeleportation(reply.teleportation);
      }

      if (hasNaN(residual)) {
        return null;
      }

      // ── Death #4 gate (Octonion Fano-plane routing) ───────────────────
      //
      // CAVEAT: Fano sum f32-parity vs canonical not yet bench-validated;
      // do not enable in prod until bench-fano-parity confirms zero
      // divergence. The gate consumption here is OPT-IN via `alphaJump`
      // and the f32-parity floor lives outside this file (the operator
      // who supplies `alphaJump.fanoSubstitute` owns that gate).
      //
      // Wire-format gap (G3): the `/residual-seed` ack does NOT carry the
      // accumulated Fano sum back to the coordinator (see
      // `apps/worker-node1/src/residual-seed-handler.ts:396-408`). The
      // `fanoSubstitute` callback is the operator's escape hatch until
      // the wire format extends. When it returns `undefined` the
      // canonical chain wins regardless of the gate — this preserves
      // the OCTONION_ROUTING_PLAN.md §5 floor guarantee.
      const isLastStation = nodeIdx === this.stations.length - 1;
      if (alphaJump && !isLastStation) {
        const dstUrl = alphaJump.resolveStationUrl(nodeIdx, prefixHash);
        if (dstUrl !== undefined) {
          // Default dispatch: Pair X (E1_E4_E5) — the skip-attention
          // shortcut from OCTONION_ROUTING_PLAN.md §4. dstLayer is
          // station N+1's first layer. nodeIdx is the source.
          const nextStation = this.stations[nodeIdx + 1]!;
          const computed = alphaJump.computeDispatch
            ? alphaJump.computeDispatch(nodeIdx, station.endLayer - 1)
            : {
                line: FanoLine.E1_E4_E5, // Pair X — skip-attention shortcut
                sign: 'pos' as const,
                dstLayer: nextStation.startLayer & 0x0f,
              };
          if (computed) {
            const { ack, skipCanonical } = await alphaJump.dispatcher.dispatchSeedHttp(
              {
                line: computed.line,
                sign: computed.sign,
                payload: residual,
                srcLayer: station.endLayer - 1,
                dstLayer: computed.dstLayer,
                tokenSeq: position >>> 0,
                dstWorkerUrl: dstUrl,
              },
              alphaJump.fetchImpl ?? fetch,
              alphaJump.requestInit,
            );
            // Strict gate: skip ONLY when (gate authorizes) AND
            // (bitset is full) AND (operator supplied a substitute that
            // returns a buffer). Any of the three failing falls back to
            // the canonical chain — the floor.
            //
            // `bitset` is not in the typed `ResidualSeedAck` (the wire
            // returns it for diagnostics) so we read it through a cast.
            const ackWithBitset = ack as (ResidualSeedAck & { bitset?: number }) | null;
            if (
              skipCanonical &&
              ackWithBitset !== null &&
              typeof ackWithBitset.bitset === 'number' &&
              ackWithBitset.bitset === FANO_FULL_MASK &&
              alphaJump.fanoSubstitute !== undefined
            ) {
              const substitute = alphaJump.fanoSubstitute(ackWithBitset, residual);
              if (substitute !== undefined) {
                if (substitute.length !== this.hiddenDim) {
                  throw new Error(
                    `Pipeline.chooseTokenThroughStations alphaJump substitute: ` +
                      `got ${substitute.length} f32, expected hiddenDim=${this.hiddenDim}`,
                  );
                }
                // Skip the next canonical hop: the substitute IS the
                // residual that station N+1 would have produced.
                residual = substitute;
                if (hasNaN(residual)) return null;
                nodeIdx++; // advance past station N+1
                traceLog(
                  `ALPHA_JUMP_SKIP node=${nodeIdx} (Fano-sum substitute, ack.bitset=0x${ackWithBitset.bitset.toString(16)})`,
                );
              }
            }
          }
        }
      }
    }

    const sampleNext = this.exitStation.sampleNext;
    if (sampleNext && this.canUseRemoteGreedySample(samplingContext)) {
      try {
        const selectedTokenId = await sampleNext.call(this.exitStation, {
          residual,
          sequenceLength: 1,
          requestId: this.requestId,
        });
        this.emitObservedTerminalProsody(
          samplingContext,
          selectedTokenId,
          'gnosis:pipeline:sample-next:observe',
        );
        return selectedTokenId;
      } catch (error) {
        if (!this.sampleNextUnavailable(error)) throw error;
      }
    }

    return await this.chooseTokenFromResidual(
      residual,
      samplingContext,
    );
  }

  private updateTerminalTailState(
    tokenId: number,
    current: TerminalTailState,
  ): TerminalTailState {
    const config = this.terminalProsody;
    if (!config) return current;
    if (config.sentenceCloseTokenIds) {
      for (const closeId of config.sentenceCloseTokenIds) {
        if (closeId === tokenId) return 'sentence-closed';
      }
    }
    const decoded = config.decodeToken?.(tokenId).trim();
    if (decoded && /^[.!?]+[\]})"'`]*$/.test(decoded)) {
      return 'sentence-closed';
    }
    return 'open-clause';
  }

  private nativeTerminalProsodyConfig():
    | {
        sentenceCloseTokenIds?: readonly number[];
        stopTokenIds?: readonly number[];
        tailState?: 'open-clause' | 'sentence-closed';
      }
    | undefined {
    const config = this.terminalProsody;
    if (!config || config.enabled === false || config.mode === 'observe') {
      return undefined;
    }
    const terminal: {
      sentenceCloseTokenIds?: readonly number[];
      stopTokenIds?: readonly number[];
      tailState?: 'open-clause' | 'sentence-closed';
    } = {};
    if (config.sentenceCloseTokenIds) {
      terminal.sentenceCloseTokenIds = Array.from(config.sentenceCloseTokenIds);
    }
    if (config.stopTokenIds) {
      terminal.stopTokenIds = Array.from(config.stopTokenIds);
    }
    if (config.tailState) {
      terminal.tailState = config.tailState;
    }
    return terminal;
  }

  /**
   * Prepare prompt prefill and report the source of the residual/logits.
   *
   * Amplituhedron coordinator (Death #2): when enabled (default ON; disable
   * with `GNOSIS_AMPLITUHEDRON_COORDINATOR=0`), the host first asks every
   * station whether it already holds a precomputed volume for this prompt
   * prefix. On an ALL-hit fan-out, the prefill phase is skipped entirely
   * and the exit station's cached tail residual seeds decode directly.
   * On ANY miss, the standard `pipelinedPrefill` runs and the resulting
   * tail residual is captured back into each station's volume cache via a
   * fire-and-forget `/amplituhedron/capture` call.
   *
   * Per `AMPLITUHEDRON_PLAN.md` §5: replay must NOT pass the tail residual
   * through `cached_mat_vec` — the residual we're seeding is itself the
   * source of truth. The host therefore feeds it directly to lm_head
   * (mirroring the tail of `pipelinedPrefill`); no station forward calls
   * fire on the prompt chunk in the all-hit branch.
   */
  async preparePrefill(args: PipelinePreparePrefillArgs): Promise<PipelinePreparedPrefill> {
    if (!args.promptTokens.length) {
      throw new Error('Pipeline.preparePrefill: promptTokens must be non-empty');
    }
    const computeLogits = args.computeLogits !== false;

    // ── Amplituhedron 4-step coordinator (default ON) ──
    //   1. prefix_hash = SHA-256-trunc-u64(promptTokens)  (blake3-substitute)
    //   2. Fan-out POST /amplituhedron/replay to every station.
    //   3. All hit → skip prefill, use the EXIT station's tail residual
    //                directly as the lm_head input; seed decode from there.
    //   4. Any miss → run pipelinedPrefill, then fan-out
    //                POST /amplituhedron/capture per station (fire-and-forget).
    if (amplituhedronCoordinatorEnabled() && this.stations.length > 0) {
      const prefixLen = args.promptTokens.length;
      const prefixHash = await prefixHashU64(args.promptTokens, this.amplituhedronModelKey);
      const replayStartedAt = Date.now();
      const replay = await this.tryAmplituhedronReplay(prefixHash, prefixLen);
      const replayMs = Date.now() - replayStartedAt;
      if (replay.hit) {
        // All stations hit. The exit station returned the post-final-layer
        // tail residual; run lm_head on it (one matmul on the exit) instead
        // of the full N-station prefill chain.
        traceLog(
          `AMPLITUHEDRON HIT prefix_hash=${prefixHash} prefix_len=${prefixLen} ` +
            `xor=${replay.xorCertificate.parityClass}`,
        );
        // Skip the lm_head matmul when the caller won't consume logits
        // (maxNewTokens === 0 / cache-warm probe): an all-hit prefix has
        // nothing to capture, so there is no work to do at all.
        const logits = computeLogits
          ? await this.lmHeadFromTail(replay.tailResidual!)
          : new Float32Array(0);
        const prefill: PrefillResult = {
          logits,
          stationCallCount: this.stations.length, // one replay call per station
          poisoned: false,
          poisonChunk: -1,
        };
        this.amplituhedronHits++;
        return {
          prefill,
          telemetry: {
            source: 'amplituhedron-hit',
            prefixHash,
            prefixLen,
            replayMs,
            prefillMs: 0,
            captureScheduled: false,
            xorStatus: replay.xorCertificate.status,
            xorDefectRank: replay.xorCertificate.defectRank,
            stationCallCount: prefill.stationCallCount,
          },
        };
      }
      traceLog(
        `AMPLITUHEDRON MISS prefix_hash=${prefixHash} prefix_len=${prefixLen} ` +
          `defect_rank=${replay.xorCertificate.defectRank} ` +
          `xor=${replay.xorCertificate.parityClass}`,
      );
      if (
        replay.partialResidual !== undefined &&
        replay.partialStartNodeIdx !== undefined
      ) {
        traceLog(
          `AMPLITUHEDRON PARTIAL prefix_hash=${prefixHash} prefix_len=${prefixLen} ` +
            `start_node=${replay.partialStartNodeIdx}`,
        );
        const prefillStartedAt = Date.now();
        const prefill = await this.prefillFromReplaySuffix({
          prefixLen,
          residual: replay.partialResidual,
          startNodeIdx: replay.partialStartNodeIdx,
          computeLogits,
          onProgress: args.onPrefillProgress,
        });
        const prefillMs = Date.now() - prefillStartedAt;
        this.scheduleAmplituhedronCapture(prefixHash, prefixLen, prefill);
        this.amplituhedronPartials++;
        return {
          prefill,
          telemetry: {
            source: 'amplituhedron-partial',
            prefixHash,
            prefixLen,
            replayMs,
            prefillMs,
            captureScheduled: true,
            partialStartNodeIdx: replay.partialStartNodeIdx,
            xorStatus: replay.xorCertificate.status,
            xorDefectRank: replay.xorCertificate.defectRank,
            stationCallCount: prefill.stationCallCount,
          },
        };
      }
      // Any miss: fall through to normal prefill, then capture. Capturing the
      // per-station prefix residuals is what lets a later partial-replay
      // recompute only the defect suffix, so it must run even when logits are
      // skipped (maxNewTokens === 0 cache-warm).
      const prefillStartedAt = Date.now();
      const prefill = await this.pipelinedPrefill({
        promptTokens: args.promptTokens,
        computeLogits,
        captureStationResiduals: true,
        onProgress: args.onPrefillProgress,
      });
      const prefillMs = Date.now() - prefillStartedAt;
      this.scheduleAmplituhedronCapture(prefixHash, prefixLen, prefill);
      this.amplituhedronMisses++;
      return {
        prefill,
        telemetry: {
          source: 'amplituhedron-miss',
          prefixHash,
          prefixLen,
          replayMs,
          prefillMs,
          captureScheduled: true,
          xorStatus: replay.xorCertificate.status,
          xorDefectRank: replay.xorCertificate.defectRank,
          stationCallCount: prefill.stationCallCount,
        },
      };
    }

    // Coordinator disabled: legacy path.
    // Phase 1: chunked pipelined prefill over the prompt.
    const prefillStartedAt = Date.now();
    const prefill = await this.pipelinedPrefill({
      promptTokens: args.promptTokens,
      computeLogits,
      onProgress: args.onPrefillProgress,
    });
    const prefillMs = Date.now() - prefillStartedAt;

    return {
      prefill,
      telemetry: {
        source: 'disabled',
        prefixLen: args.promptTokens.length,
        replayMs: 0,
        prefillMs,
        captureScheduled: false,
        stationCallCount: prefill.stationCallCount,
      },
    };
  }

  /**
   * Shared-prefix KV reuse (token axis).
   *
   * The Amplituhedron coordinator (`preparePrefill`) keys its volume cache on
   * the hash of the WHOLE prompt, so it only ever hits on an *identical* prompt.
   * For a workload where many prompts share a common LEADING run of tokens — a
   * fixed chat-template preamble like `<|im_start|>user\n`, a shared system
   * prompt, or the carried context of a multi-turn conversation — that full-prompt
   * key always misses on the divergent suffix and the shared prefix is re-prefilled
   * every single request.
   *
   * This method captures/replays the shared prefix as its OWN Amplituhedron key
   * (`prefixHash(promptTokens[0..sharedPrefixLen)), prefixLen=sharedPrefixLen`).
   * On a replay HIT every station's `set_kv_slab` reseeds the KV cache for token
   * positions `[0, sharedPrefixLen)`, so we only have to prefill the SUFFIX tokens
   * `[sharedPrefixLen, P)` at absolute `startPosOffset = sharedPrefixLen` — the
   * suffix forward attends causally to the reseeded prefix KV. The shared prefill
   * (the `sharedPrefixLen / P` fraction of the prompt) is skipped.
   *
   * HONEST scope of the win:
   *   - Distinct one-shot questions that share only a few template tokens
   *     (`sharedPrefixLen` ≈ 3): the saved fraction is tiny, so the win is small
   *     and can be net-zero once the extra replay round-trip is paid.
   *   - Multi-turn / shared-system-prompt: `sharedPrefixLen` is most of the prompt,
   *     so the saved fraction (and the TPS win) is large.
   *
   * It degrades to `preparePrefill` (the full-prompt path) whenever it cannot win:
   *   - `sharedPrefixLen <= 0` or `>= promptTokens.length` (nothing to skip / no
   *     suffix to seed decode from),
   *   - the Amplituhedron coordinator is disabled,
   *   - the shared-prefix replay misses (first request for that prefix — which
   *     then captures the shared prefix for subsequent requests),
   * so a caller can ALWAYS route through here without a correctness or floor risk.
   *
   * On a miss the FULL prompt is prefilled (via `preparePrefill`, which also
   * captures the full-prompt key) AND a second fire-and-forget capture stores the
   * shared-prefix slab (read off the station's now-warm KV via `get_kv_slab` at
   * `prefix_len = sharedPrefixLen`). Best-effort: a dropped capture only costs the
   * next same-prefix request a re-prefill — never a wrong token.
   */
  async prepareSharedPrefixPrefill(args: {
    promptTokens: Uint32Array;
    sharedPrefixLen: number;
    onPrefillProgress?: (progress: PipelinePrefillProgress) => void;
    computeLogits?: boolean;
  }): Promise<PipelinePreparedPrefill> {
    if (!args.promptTokens.length) {
      throw new Error('Pipeline.prepareSharedPrefixPrefill: promptTokens must be non-empty');
    }
    const computeLogits = args.computeLogits !== false;
    const promptLen = args.promptTokens.length;
    const sharedLen = args.sharedPrefixLen;
    // No usable shared prefix (nothing to skip, or it covers the whole prompt so
    // there is no suffix to seed decode logits from) ⇒ defer to the full-prompt
    // Amplituhedron path unchanged.
    if (
      !Number.isInteger(sharedLen) ||
      sharedLen <= 0 ||
      sharedLen >= promptLen ||
      !amplituhedronCoordinatorEnabled()
    ) {
      return this.preparePrefill({
        promptTokens: args.promptTokens,
        ...(args.onPrefillProgress ? { onPrefillProgress: args.onPrefillProgress } : {}),
        computeLogits,
      });
    }

    const sharedTokens = args.promptTokens.subarray(0, sharedLen);
    const sharedHash = await prefixHashU64(sharedTokens, this.amplituhedronModelKey);
    const replayStartedAt = Date.now();
    const replay = await this.tryAmplituhedronReplay(sharedHash, sharedLen);
    const replayMs = Date.now() - replayStartedAt;

    if (replay.hit) {
      // Every station's KV now holds positions [0, sharedLen). Prefill ONLY the
      // suffix tokens at absolute positions [sharedLen, promptLen); the suffix
      // attends causally to the reseeded prefix KV.
      const suffixTokens = args.promptTokens.slice(sharedLen, promptLen);
      const prefillStartedAt = Date.now();
      const prefill = await this.pipelinedPrefill({
        promptTokens: suffixTokens,
        startPosOffset: sharedLen,
        computeLogits,
        captureStationResiduals: true,
        ...(args.onPrefillProgress ? { onProgress: args.onPrefillProgress } : {}),
      });
      const prefillMs = Date.now() - prefillStartedAt;
      // Refresh the full-prompt key (so an exact repeat takes the all-hit fast
      // path) AND the shared-prefix key (keeps it warm for sibling suffixes).
      const fullHash = await prefixHashU64(args.promptTokens, this.amplituhedronModelKey);
      this.scheduleAmplituhedronCapture(fullHash, promptLen, prefill);
      this.scheduleSharedPrefixCapture(sharedHash, sharedLen);
      traceLog(
        `SHARED_PREFIX HIT shared_hash=${sharedHash} shared_len=${sharedLen} ` +
          `suffix_len=${suffixTokens.length} prefill_ms=${prefillMs}`,
      );
      return {
        prefill,
        telemetry: {
          source: 'amplituhedron-partial',
          prefixHash: sharedHash,
          prefixLen: sharedLen,
          replayMs,
          prefillMs,
          captureScheduled: true,
          partialStartNodeIdx: 0,
          xorStatus: replay.xorCertificate.status,
          xorDefectRank: replay.xorCertificate.defectRank,
          stationCallCount: prefill.stationCallCount,
        },
      };
    }

    // Shared-prefix MISS: full-prompt path (which captures the full key), then a
    // second capture for the shared-prefix key so later sibling suffixes hit.
    const prepared = await this.preparePrefill({
      promptTokens: args.promptTokens,
      ...(args.onPrefillProgress ? { onPrefillProgress: args.onPrefillProgress } : {}),
      computeLogits,
    });
    if (!prepared.prefill.poisoned) {
      this.scheduleSharedPrefixCapture(sharedHash, sharedLen);
    }
    traceLog(
      `SHARED_PREFIX MISS shared_hash=${sharedHash} shared_len=${sharedLen} ` +
        `source=${prepared.telemetry.source}`,
    );
    return prepared;
  }

  /**
   * Fire-and-forget capture of the shared-prefix KV slab. Mirrors
   * `scheduleAmplituhedronCapture` but keys on the shared-prefix tuple and at
   * `prefix_len = sharedLen` — the station reads its OWN current KV via
   * `get_kv_slab(layerLo, layerHi, sharedLen)`, which after a full or
   * shared-aware prefill holds the shared-prefix positions. Never awaited; a
   * dropped capture only costs the next sibling-suffix request a re-prefill.
   *
   * A hiddenDim-sized zero tail residual is passed because the host
   * `Station.amplituhedronCapture` short-circuits to a NOOP when tailResidual is
   * absent (it requires a Float32Array for the wire codec), so omitting it would
   * silently never reach the station. The tail value itself is irrelevant for a
   * shared prefix: on replay we consume only the reseeded KV slab (`set_kv_slab`)
   * and recompute the residual by forwarding the divergent suffix — we never use
   * the shared prefix's `set_batch_x` tail. The station re-derives the real KV
   * slab from its own warm state (`get_kv_slab(layerLo, layerHi, sharedLen)`).
   */
  private scheduleSharedPrefixCapture(sharedHash: string, sharedLen: number): void {
    const zeroTail = new Float32Array(this.hiddenDim);
    for (let stationIndex = 0; stationIndex < this.stations.length; stationIndex++) {
      const station = this.stations[stationIndex]!;
      const capable = asAmplituhedronCapable(station);
      if (!capable.amplituhedronCapture) continue;
      fireAndForget(
        capable.amplituhedronCapture({
          prefixHash: sharedHash,
          prefixLen: sharedLen,
          layerLo: station.startLayer,
          layerHi: station.endLayer,
          tailResidual: zeroTail,
          requestId: this.requestId,
        }),
        `shared-prefix capture layer=${station.startLayer}-${station.endLayer}`,
      );
    }
  }

  /**
   * Amplituhedron volume-cache hit/miss tallies for this process lifetime (+ the chunk
   * cache), exposed via GET /cache/stats. This is the live `cache-hit-rate` that
   * docs/COST_MODEL.md needs to turn the blended substrate cost from an assumed point into
   * a measured one. Counts only — no keys, no payloads. Partials count as half a hit (they
   * replay a prefix and recompute only the defect suffix).
   */
  getCacheStats(): {
    amplituhedron: {
      hits: number;
      partials: number;
      misses: number;
      total: number;
      hitRate: number;
    };
    chunk: { hits: number; misses: number; total: number; hitRate: number };
  } {
    const aTotal =
      this.amplituhedronHits + this.amplituhedronPartials + this.amplituhedronMisses;
    const aHitEquiv = this.amplituhedronHits + 0.5 * this.amplituhedronPartials;
    const cTotal = this.chunkCacheHits + this.chunkCacheMisses;
    return {
      amplituhedron: {
        hits: this.amplituhedronHits,
        partials: this.amplituhedronPartials,
        misses: this.amplituhedronMisses,
        total: aTotal,
        hitRate: aTotal > 0 ? aHitEquiv / aTotal : 0,
      },
      chunk: {
        hits: this.chunkCacheHits,
        misses: this.chunkCacheMisses,
        total: cTotal,
        hitRate: cTotal > 0 ? this.chunkCacheHits / cTotal : 0,
      },
    };
  }

  /**
   * Wallington Rotation prefill + autoregressive generate.
   * Returns `[...promptTokens, ...newTokens]`.
   */
  async generate(args: {
    promptTokens: Uint32Array;
    maxNewTokens: number;
    onPrefillProgress?: (progress: PipelinePrefillProgress) => void;
    onToken?: (event: PipelineTokenEvent) => void;
    suppressTerminalProsody?: boolean;
    /**
     * Death #4 opt-in. When provided, the per-token canonical chain
     * consults the supplied `AlphaJumpDispatcher` after each station hop;
     * see `AlphaJumpHook` for the full contract and the Fano-sum CAVEAT.
     * Default undefined → behavior unchanged from pre-Death-#4.
     */
    alphaJump?: AlphaJumpHook;
  }): Promise<Uint32Array> {
    if (!args.promptTokens.length) {
      throw new Error('Pipeline.generate: promptTokens must be non-empty');
    }
    if (!Number.isInteger(args.maxNewTokens) || args.maxNewTokens < 0) {
      throw new Error(
        `Pipeline.generate: maxNewTokens must be a non-negative integer (got ${String(args.maxNewTokens)})`,
      );
    }

    const nativeGenerate = this.entryStation.generateTokens;
    if (
      this.stations.length === 1 &&
      nativeGenerate &&
      !this.disableNativeGenerate &&
      !args.alphaJump &&
      // Personality steering applies the per-token bias on the `/decode-next`
      // path only; the native `/generate` fast path does not, so skip it when
      // steering is active for this request.
      !this.steeringActive &&
      // Speculative decode lives in generateFromPrefill (the host-orchestrated
      // decode below), NOT the native single-call /generate. When spec-decode is
      // enabled and eligible, bypass native /generate so the spec loop runs.
      !this.shouldSpeculate(args.alphaJump !== undefined)
    ) {
      const nativeArgs: Parameters<NonNullable<typeof nativeGenerate>>[0] = {
        tokens: args.promptTokens,
        maxNewTokens: args.maxNewTokens,
      };
      if (this.requestId !== undefined) {
        nativeArgs.requestId = this.requestId;
      }
      const terminalProsody = args.suppressTerminalProsody
        ? undefined
        : this.nativeTerminalProsodyConfig();
      if (terminalProsody !== undefined) {
        nativeArgs.terminalProsody = terminalProsody;
      }
      try {
        const generated = await nativeGenerate.call(this.entryStation, {
          ...nativeArgs,
        });
        const out = new Uint32Array(args.promptTokens.length + generated.length);
        out.set(args.promptTokens);
        out.set(generated, args.promptTokens.length);
        for (let i = 0; i < generated.length; i++) {
          args.onToken?.({
            tokenId: generated[i]!,
            tokenIndex: i,
            maxNewTokens: args.maxNewTokens,
          });
        }
        return out;
      } catch (error) {
        if (!this.nativeGenerateUnavailable(error)) throw error;
      }
    }

    // Prefill (Amplituhedron coordinator + telemetry) is delegated to
    // preparePrefill; generate() owns only the native fast path above and the
    // autoregressive decode below.
    const prepared = await this.preparePrefill({
      promptTokens: args.promptTokens,
      onPrefillProgress: args.onPrefillProgress,
      computeLogits: args.maxNewTokens > 0,
    });

    let prefixHash: string | undefined = undefined;
    if (args.alphaJump) {
      prefixHash = await prefixHashU64(args.promptTokens);
    }

    if (args.maxNewTokens === 0) {
      return new Uint32Array(args.promptTokens);
    }

    return await this.generateFromPrefill({
      promptTokens: args.promptTokens,
      prefixHash,
      prefill: prepared.prefill,
      maxNewTokens: args.maxNewTokens,
      onToken: args.onToken,
      alphaJump: args.alphaJump,
    });
  }

  /**
   * SEMANTIC-KEY ACCESSOR (sovereign sieve — semantic-sieve-cache.ts).
   *
   * Return the prompt's final pre-lm_head hidden — the transformer's
   * post-all-layers tail residual of the last prompt token, shape [hiddenDim] —
   * as the resonance key for the semantic cache, or `null` on a capability miss.
   *
   * REACHABILITY (the honest contract):
   *   - This drives the HOST-ORCHESTRATED prefill (`preparePrefill`), which is the
   *     ONLY path that returns the tail residual to the host. The native
   *     single-station `/generate` fast path computes the whole forward INSIDE the
   *     Rust fat-station and returns ONLY token ids, so its tail hidden never
   *     crosses the wire — that is why we call preparePrefill directly here rather
   *     than reading it off `generate()`.
   *   - The prefill this runs WARMS the Amplituhedron prefix cache, so the
   *     subsequent `generate()`'s own preparePrefill replays it (all-hit) cheaply —
   *     the two prefills are largely amortized by the existing cache, not 2× compute.
   *   - Returns `null` when the tail residual is not produced: (a) alt-arch
   *     `/generate`-only knots (gemma/mamba/rwkv7/qwen35moe) whose `/embed`+`/forward`
   *     host path is unsupported — callers MUST gate this off for alt-arch; and
   *     (b) an Amplituhedron ALL-HIT prefix, where the tail residual is consumed for
   *     lm_head but not surfaced in PrefillResult (already the fast exact path, so
   *     the semantic layer adds nothing there).
   *
   * Callers MUST treat `null` as "no semantic key" and fall through to normal
   * decode WITHOUT seeding.
   *
   * STATUS (the fold landed): the native `/generate` accessor now exists —
   * `Pipeline.generateWithPromptHidden` issues ONE fat-station `/generate` with
   * `return_hidden: true`, so a single forward produces BOTH the semantic key and
   * the first-K decode tokens (see fat-station.rs commit 02618367). This method
   * stays as the DOCUMENTED FALLBACK for the multi-station / host-orchestrated
   * transformer path (where the native single-call generate is not used), and for
   * a native knot that does not surface `prompt_hidden`. `null` semantics unchanged.
   */
  async computePromptHidden(promptTokens: Uint32Array): Promise<Float32Array | null> {
    if (!promptTokens.length) return null;
    const prepared = await this.preparePrefill({ promptTokens, computeLogits: false });
    return prepared.prefill.tailResidual ?? null;
  }

  /**
   * THE FOLD — one forward serves BOTH the sovereign-sieve semantic KEY and the
   * ANSWER (its first `maxNewTokens` decode tokens).
   *
   * On the native single-station fast path this issues ONE fat-station `/generate`
   * with `return_hidden: true` (see fat-station.rs `prompt_hidden`, commit 02618367):
   * the response carries `generated_tokens` (the first-K decode tokens for the
   * caller's decode-agreement gate) AND `promptHidden` (the last prompt token's tail
   * residual = the semantic key). No separate host-orchestrated prefill (the old
   * `computePromptHidden` round-trip) and no separate probe generate — the key comes
   * from the prefill this call runs anyway.
   *
   * FALLBACK (multi-station / host-orchestrated transformer path, or a native knot
   * that omits `prompt_hidden`): key via `computePromptHidden` (host-orchestrated
   * prefill) + a normal `generate`. Two prefills there, but the Amplituhedron prefix
   * cache amortizes the second — the same contract `computePromptHidden` documents.
   *
   * `promptHidden` is null on a full capability miss; the caller MUST treat null as
   * "no semantic key" (skip the sieve, do NOT seed) and fall through to normal decode.
   */
  async generateWithPromptHidden(args: {
    promptTokens: Uint32Array;
    maxNewTokens: number;
  }): Promise<{ tokens: Uint32Array; promptHidden: Float32Array | null }> {
    if (!args.promptTokens.length) {
      throw new Error('Pipeline.generateWithPromptHidden: promptTokens must be non-empty');
    }
    if (!Number.isInteger(args.maxNewTokens) || args.maxNewTokens < 0) {
      throw new Error(
        `Pipeline.generateWithPromptHidden: maxNewTokens must be a non-negative integer (got ${String(args.maxNewTokens)})`,
      );
    }
    const nativeWithHidden = this.entryStation.generateTokensWithHidden;
    if (
      this.stations.length === 1 &&
      nativeWithHidden &&
      !this.disableNativeGenerate &&
      // Steering forces the /decode-next per-token path (no native /generate), so the
      // fold is unavailable there — fall through to the host-orchestrated fallback.
      !this.steeringActive
    ) {
      const nativeArgs: Parameters<NonNullable<typeof nativeWithHidden>>[0] = {
        tokens: args.promptTokens,
        maxNewTokens: args.maxNewTokens,
      };
      if (this.requestId !== undefined) {
        nativeArgs.requestId = this.requestId;
      }
      try {
        const { tokens: generated, promptHidden } = await nativeWithHidden.call(
          this.entryStation,
          nativeArgs,
        );
        const out = new Uint32Array(args.promptTokens.length + generated.length);
        out.set(args.promptTokens);
        out.set(generated, args.promptTokens.length);
        logSievePromptHiddenPath('native', this.stations.length, promptHidden);
        return { tokens: out, promptHidden };
      } catch (error) {
        // A knot without the native `/generate` accessor falls back to the
        // host-orchestrated path below; any other error propagates.
        if (!this.nativeGenerateUnavailable(error)) throw error;
      }
    }
    // Fallback: host-orchestrated key + normal generate (see docstring). The key is
    // best-effort — a null key just disables the sieve for this request.
    const promptHidden = await this.computePromptHidden(args.promptTokens).catch(() => null);
    const tokens = await this.generate({
      promptTokens: args.promptTokens,
      maxNewTokens: args.maxNewTokens,
      suppressTerminalProsody: true,
    });
    logSievePromptHiddenPath('fallback', this.stations.length, promptHidden);
    return { tokens, promptHidden };
  }

  /**
   * Install personality logit-steering on the exit station for the duration of a
   * request (per-request; the pipeline is a shared singleton). Sets a flag that
   * forces the autoregressive `/decode-next` path where the bias is applied.
   * The caller MUST `clearPersonalitySteering()` in a `finally` so steering
   * never leaks into the next request on the same station.
   */
  async installPersonalitySteering(spec: PersonalitySteeringSpec): Promise<void> {
    const install = this.exitStation.installPersonalitySteering;
    if (!install) return;
    await install.call(this.exitStation, spec, this.requestId);
    this.steeringActive = true;
  }

  /** Best-effort clear; never fails the response on cleanup error. */
  async clearPersonalitySteering(): Promise<void> {
    this.steeringActive = false;
    const clear = this.exitStation.clearPersonalitySteering;
    if (!clear) return;
    try {
      await clear.call(this.exitStation, this.requestId);
    } catch (error) {
      traceLog(`personality steering clear failed: ${String(error)}`);
    }
  }

  /**
   * Fan out `/amplituhedron/replay` to every station in parallel. Returns
   * the exit station's tail residual on ALL-hit; `null` on any miss or any
   * station that doesn't implement the optional method.
   *
   * The exit station's residual is the only one the coordinator needs —
   * mid-pipeline stations participate by reseeding their internal KV
   * slabs (the load-bearing side effect on their `set_kv_slab` call);
   * their tail residual is irrelevant to the host because the next call
   * on the prompt chain is lm_head, not another forward.
   */
  private async tryAmplituhedronReplay(
    prefixHash: string,
    prefixLen: number,
  ): Promise<AmplituhedronReplayDecision> {
    const stationWindow = this.resolveAmplituhedronStationWindow(prefixHash);
    const replayPromises: Promise<AmplituhedronReplayResult>[] = [];
    for (
      let stationIndex = stationWindow.start;
      stationIndex < stationWindow.end;
      stationIndex++
    ) {
      const station = this.stations[stationIndex]!;
      const capable = asAmplituhedronCapable(station);
      if (!capable.amplituhedronReplay) {
        replayPromises.push(Promise.resolve({ hit: false } as AmplituhedronReplayResult));
        continue;
      }
      replayPromises.push(
        capable.amplituhedronReplay({
          prefixHash,
          prefixLen,
          layerLo: station.startLayer,
          layerHi: station.endLayer,
          requestId: this.requestId,
        }).catch((err) => {
          traceLog(
            `AMPLITUHEDRON replay error layer=${station.startLayer}-${station.endLayer}: ${String(err)}`,
          );
          return { hit: false } as AmplituhedronReplayResult;
        }),
      );
    }
    const replies = await Promise.all(replayPromises);
    const xorCertificate = issueAmplituhedronReplayXorCertificate({
      prefixHash,
      prefixLen,
      stationHits: replies.map((reply) => reply.hit),
    });
    if (!replies.every((r) => r.hit)) {
      const firstMissIndex = replies.findIndex((reply) => !reply.hit);
      if (stationWindow.sharded) {
        return { hit: false, xorCertificate };
      }
      const seedIndex = firstMissIndex - 1;
      const seed =
        seedIndex >= 0 ? replies[seedIndex]?.prefixResidual : undefined;
      if (seed && seed.length === prefixLen * this.hiddenDim) {
        return {
          hit: false,
          partialStartNodeIdx: firstMissIndex,
          partialResidual: seed,
          partialFingerprint: replies[seedIndex]?.prefixFingerprint,
          xorCertificate,
        };
      }
      if (seed && seed.length !== prefixLen * this.hiddenDim) {
        traceLog(
          `AMPLITUHEDRON partial seed length mismatch node=${seedIndex} ` +
            `len=${seed.length} expected=${prefixLen * this.hiddenDim}`,
        );
      }
      return { hit: false, xorCertificate };
    }
    // Exit station's tail residual is the seed for lm_head.
    const exitReply = replies[replies.length - 1]!;
    if (!exitReply.tailResidual || exitReply.tailResidual.length !== this.hiddenDim) {
      traceLog(
        `AMPLITUHEDRON exit station hit without tail residual (len=${
          exitReply.tailResidual?.length ?? 0
        } expected=${this.hiddenDim}) — falling back to prefill`,
      );
      return {
        hit: false,
        xorCertificate: issueAmplituhedronReplayXorCertificate({
          prefixHash,
          prefixLen,
          stationHits: this.stations.map((_, index) => index !== this.stations.length - 1),
          roundSuffix: 'exit-tail-invalid',
        }),
      };
    }
    return { hit: true, tailResidual: exitReply.tailResidual, xorCertificate };
  }

  /**
   * lm_head over the replayed tail residual. Mirrors the tail of
   * `pipelinedPrefill`: cobordism-sharded comultiplication when cobordism
   * stations are configured, single exit-station call otherwise.
   *
   * Per §5: this MUST NOT route through `cached_mat_vec` (the matmul memo).
   * The host's lm_head call hits the exit station directly; the station's
   * lm_head handler in `apps/distributed-inference-worker/src/index.ts`
   * does not invoke the AmplituhedronCache — only `forward_range_*` does —
   * so the double-count is structurally prevented at the wire layer.
   */
  private async lmHeadFromTail(tailResidual: Float32Array): Promise<Float32Array> {
    if (this.cobordismStations && this.cobordismStations.length > 0) {
      const partials = await Promise.all(
        this.cobordismStations.map((cs) =>
          cs.station.cobordismLmHead({
            residual: tailResidual,
            sequenceLength: 1,
            vocabStart: cs.vocabStart,
            vocabEnd: cs.vocabEnd,
            requestId: this.requestId,
          }),
        ),
      );
      const totalVocab = this.cobordismStations.reduce(
        (acc, cs) => Math.max(acc, cs.vocabEnd),
        0,
      );
      const logits = new Float32Array(totalVocab);
      for (let i = 0; i < this.cobordismStations.length; i++) {
        const cs = this.cobordismStations[i]!;
        const partial = partials[i]!;
        if (partial.length !== cs.vocabEnd - cs.vocabStart) {
          throw new Error(
            `Pipeline.lmHeadFromTail cobordism: shard ${i} returned ${partial.length} floats, ` +
              `expected ${cs.vocabEnd - cs.vocabStart}`,
          );
        }
        logits.set(partial, cs.vocabStart);
      }
      return logits;
    }
    return await this.exitStation.lmHead({
      residual: tailResidual,
      sequenceLength: 1,
      requestId: this.requestId,
    });
  }

  private async prefillFromReplaySuffix(args: {
    prefixLen: number;
    residual: Float32Array;
    startNodeIdx: number;
    computeLogits?: boolean;
    onProgress?: (progress: PipelinePrefillProgress) => void;
  }): Promise<PrefillResult> {
    if (!Number.isInteger(args.prefixLen) || args.prefixLen <= 0) {
      throw new Error(
        `Pipeline.prefillFromReplaySuffix: prefixLen must be positive (got ${String(args.prefixLen)})`,
      );
    }
    if (
      !Number.isInteger(args.startNodeIdx) ||
      args.startNodeIdx <= 0 ||
      args.startNodeIdx >= this.stations.length
    ) {
      throw new Error(
        `Pipeline.prefillFromReplaySuffix: startNodeIdx must be inside the station suffix (got ${String(args.startNodeIdx)})`,
      );
    }
    const expectedResidualLen = args.prefixLen * this.hiddenDim;
    if (args.residual.length !== expectedResidualLen) {
      throw new Error(
        `Pipeline.prefillFromReplaySuffix: residual length ${args.residual.length} != ${expectedResidualLen}`,
      );
    }

    const batchCap = Math.min(...this.stations.map((s) => s.maxBatchSize ?? 4));
    const chunkSize = Math.max(1, Math.min(batchCap, this.explicitChunkSize ?? batchCap));
    const chunks: ChunkDescriptor[] = [];
    for (let pos = 0; pos < args.prefixLen; pos += chunkSize) {
      chunks.push({ startPos: pos, count: Math.min(chunkSize, args.prefixLen - pos) });
    }

    let stationCallCount = 0;
    let poisonChunk = -1;
    let lastHiddenStates: Float32Array<ArrayBufferLike> = new Float32Array(this.hiddenDim);
    args.onProgress?.({
      completedTokens: 0,
      totalTokens: args.prefixLen,
      completedChunks: 0,
      totalChunks: chunks.length,
    });

    for (let chunkIdx = 0; chunkIdx < chunks.length; chunkIdx++) {
      const chunk = chunks[chunkIdx]!;
      const offset = chunk.startPos * this.hiddenDim;
      let residual = copyView(
        args.residual.subarray(offset, offset + chunk.count * this.hiddenDim),
      );

      for (let nodeIdx = args.startNodeIdx; nodeIdx < this.stations.length; nodeIdx++) {
        const station = this.stations[nodeIdx]!;
        stationCallCount++;
        const reply = await station.forward({
          residual,
          startPos: chunk.startPos,
          sequenceLength: chunk.count,
          requestId: this.requestId,
        });
        residual = reply.residual;
        if (this.enableTeleportation && this.teleportationOptimizer && reply.teleportation) {
          this.updateTeleportation(reply.teleportation);
        }
        if (hasNaN(residual)) {
          poisonChunk = chunkIdx;
          break;
        }
      }

      if (poisonChunk < 0) {
        lastHiddenStates = residual;
      }
      args.onProgress?.({
        completedTokens: Math.min(args.prefixLen, chunk.startPos + chunk.count),
        totalTokens: args.prefixLen,
        completedChunks: chunkIdx + 1,
        totalChunks: chunks.length,
      });
      if (poisonChunk >= 0) break;
    }

    const lastChunk = chunks[chunks.length - 1]!;
    const tailResidual =
      lastHiddenStates.length === this.hiddenDim
        ? lastHiddenStates
        : lastHiddenStates.subarray(
            (lastChunk.count - 1) * this.hiddenDim,
            lastChunk.count * this.hiddenDim,
          );
    const logits =
      args.computeLogits === false
        ? new Float32Array(0)
        : await this.lmHeadFromTail(tailResidual);
    if (args.computeLogits !== false) stationCallCount++;
    return {
      logits,
      stationCallCount,
      poisoned: poisonChunk >= 0,
      poisonChunk,
      tailResidual: copyView(tailResidual),
    };
  }

  /**
   * Fire-and-forget `/amplituhedron/capture` per station after a miss-path
   * prefill completes. The station's wasm side already holds the freshly
   * computed KV slab and tail residual (`get_kv_slab` / `get_residual`),
   * so the capture body only needs to carry the key tuple — the wire
   * handler in `apps/distributed-inference-worker/src/index.ts` reads the
   * tensors directly off the pipeline.
   *
   * Capture is NEVER awaited. Per the plan, the cost is one slab-copy on
   * the station; deferring it to a CF Worker `waitUntil` keeps it off the
   * decode-loop critical path.
   */
  private scheduleAmplituhedronCapture(
    prefixHash: string,
    prefixLen: number,
    prefill: PrefillResult,
  ): void {
    const stationWindow = this.resolveAmplituhedronStationWindow(prefixHash);

    for (
      let stationIndex = stationWindow.start;
      stationIndex < stationWindow.end;
      stationIndex++
    ) {
      const station = this.stations[stationIndex]!;
      const capable = asAmplituhedronCapable(station);
      if (!capable.amplituhedronCapture) continue;
      const prefixResidual = prefill.stationPrefixResiduals?.[stationIndex];
      fireAndForget(
        capable.amplituhedronCapture({
          prefixHash,
          prefixLen,
          layerLo: station.startLayer,
          layerHi: station.endLayer,
          tailResidual: prefixResidual ?? prefill.tailResidual,
          requestId: this.requestId,
        }),
        `capture layer=${station.startLayer}-${station.endLayer}`,
      );
    }
  }

  private verifyBatchUnavailable(error: unknown): boolean {
    const message = error instanceof Error ? error.message : String(error);
    return (
      message.includes('/verify-batch') &&
      (message.includes('HTTP 404') || message.includes('not available'))
    );
  }

  /**
   * Whether the greedy prompt-lookup / n-gram speculative-decode path is
   * eligible for this request. ALL must hold (else the path is not provably
   * lossless or not single-batch-verifiable, so we use the baseline loop):
   *   - the flag is on (config.specDecode ?? MESH_SPEC_DECODE env)
   *   - exactly one station, and it exposes the `/verify-batch` primitive
   *   - the default greedy argmax sampler (no custom sampler)
   *   - no personality steering installed for this request
   *   - no terminal-prosody cadence biasing configured
   *   - no per-token AlphaJump dispatcher
   */
  private shouldSpeculate(hasAlphaJump: boolean): boolean {
    return (
      this.specDecode &&
      this.stations.length === 1 &&
      typeof this.entryStation.verifyBatch === 'function' &&
      this.usesDefaultSampler &&
      !this.steeringActive &&
      // The openai-server ALWAYS constructs the terminal-prosody *capability*
      // config (per-request opt-in), so `terminalProsody === undefined` is never
      // true and would dead-gate spec. The lossless condition is that prosody is
      // not actively BIASING token selection: when bias is inactive (observe/off
      // or suppressed) the baseline is pure greedy argmax, identical to the spec
      // loop's emitted tokens. Bias-active (terminal-window cadence) excludes it.
      !this.terminalProsodyBiasActive() &&
      !hasAlphaJump
    );
  }

  /**
   * GREEDY prompt-lookup / n-gram speculative decode loop (LOSSLESS).
   *
   * State after prefill (the part the original implementation got wrong):
   * the host's prefill path is `/forward` (returns a residual; it does NOT
   * write `self.logits`) + `/lm-head` (writes `self.batch_logits`, NOT
   * `self.logits`). So after prefill the entry station's `self.pos` is NOT set
   * (the `/forward` path advances the KV but the station's `pos` field is
   * driven by `X-Position`, not stored) and `self.logits` is STALE. The
   * next-token distribution for the slot after the prompt lives ONLY in
   * `args.prefill.logits`, which the HOST holds. Therefore we cannot let the
   * first `/verify-batch` round read the station's logits for its first token.
   *
   * Seed: the host emits the first token itself — exactly like the baseline
   * loop — as `chooseToken(args.prefill.logits)` (pure argmax; spec is gated on
   * the default sampler with no active prosody bias), then commits it via
   * `/decode-next` at the prompt-end slot. `/decode-next` runs a real
   * `forward_one`, which (a) writes the committed token's KV and (b) leaves
   * `self.logits` holding the genuine model distribution for the NEXT slot, and
   * (c) sets `self.pos = promptLen + 1`. After the seed the station's
   * `self.logits` is valid, so every subsequent `/verify-batch` round's first
   * candidate (`i == 0`, which reads `original_logits = self.logits`) and the
   * empty-candidate fast path are correct.
   *
   * Then each round (host tracks the running `seq` = prompt + generated):
   *
   *   draft      = ngramDraft(seq)            // predicts slots seq.len()...
   *   candidates = draft.slice(0, budget)     // may be EMPTY (no n-gram match)
   *   { accepted, nextToken } = verifyBatch(candidates, position: seq.length)
   *
   * `position` is the ABSOLUTE KV slot of the first candidate. The fat-station
   * seeds `self.pos = position` before the batched candidate forward, so the
   * candidate KV is written at the correct slots and the warm prompt KV is never
   * clobbered (the bug: `self.pos` defaulted stale-0 → candidates written at
   * slot 0 → context-free garbage with clustered-repeating tokens). After the
   * seed, `seq.length == self.pos` every round, so `position` and the station
   * agree.
   *
   * `/verify-batch` forwards the candidate span in ONE batched forward, accepts
   * the longest argmax-matching prefix, commits `accepted ++ [nextToken]` to KV,
   * and returns them. `nextToken` is ALWAYS produced (the station's own argmax
   * for the slot after the accepted prefix), so an EMPTY candidate set degrades
   * to exactly one normal greedy step (accepted == []). We emit
   * `accepted ++ [nextToken]` and advance `seq`.
   *
   * LOSSLESS: the seed token is `argmax(prefill.logits)` (identical to baseline
   * line "chooseToken(args.prefill.logits)"); every subsequent emitted token is
   * the model's argmax at its slot (the `inner_verify_batch` accept rule for
   * accepted tokens; the post-verify argmax for `nextToken`), computed against a
   * correctly-positioned KV. Identical to the baseline greedy loop token-for-token.
   *
   * Returns the completed token array, or `null` if the station turns out not to
   * support `/verify-batch` OR `/decode-next` (404 / not-available) BEFORE any
   * token is emitted — the caller then runs the untouched baseline loop. Stop
   * tokens and `maxNewTokens` are honored: emitted tokens are truncated at the
   * budget and emission halts at the first stop token (matching the baseline
   * loop's early-return semantics).
   */
  private async generateFromPrefillSpeculative(
    args: PipelineGenerateFromPrefillArgs,
    out: Uint32Array,
    startOutLen: number,
  ): Promise<Uint32Array | null> {
    const verifyBatch = this.entryStation.verifyBatch;
    const decodeNext = this.entryStation.decodeNext;
    // Spec needs BOTH primitives: verify-batch for the drafted rounds and
    // decode-next to SEED the first token (the prefill leaves `self.logits`
    // stale, so the station cannot produce the first token itself). If either is
    // missing, bail to the baseline loop on the untouched station.
    if (!verifyBatch || !decodeNext) return null;

    let outLen = startOutLen;
    const promptLen = args.promptTokens.length;
    const maxNew = args.maxNewTokens;
    const finish = (): Uint32Array =>
      outLen === out.length ? out : out.slice(0, outLen);

    const generatedTokens: number[] = [];
    let terminalTailState = this.terminalProsody?.tailState ?? 'open-clause';

    // Running token sequence (prompt + generated), used by the n-gram draft.
    const seq: number[] = Array.from(args.promptTokens);
    const kCap = this.specDecodeK;

    // ── Self-calibrating adaptive cost-gate (analogous to gnosis-frf's certified
    // cost gate). The break-even derivation (live L4 measurements are ground
    // truth): a single-token verify (candidates=[]) costs t1≈76ms; a K-candidate
    // verify costs tK≈242ms for K=4, so R = tK/t1 ≈ 3.2. A DRAFTED round emits
    // (a+1) tokens for cost tK; a 1-TOKEN round emits 1 for cost t1. Per-token
    // cost is tK/(a+1) vs t1, so drafting wins iff tK/(E[a]+1) < t1, i.e.
    //   E[a] > R - 1   (≈ 2.2 for R=3.2).
    // We can't know E[a] a priori, so we ESTIMATE it from a rolling
    // exponentially-weighted acceptance rate and only draft while the estimate
    // clears the break-even. This adapts WITHIN one generation: a repetitive run
    // (JSON/code) pushes the EWMA up and we draft aggressively at full K; a prose
    // / counting stretch drops it below break-even and we BACK OFF to cheap
    // single-token steps, then PROBE occasionally to recover when text repeats
    // again. K itself shrinks on low acceptance (cheaper wrong drafts) and grows
    // back toward the cap on high acceptance. Lossless throughout: gating only
    // decides WHETHER / HOW MANY candidates to draft — candidates=[] is always an
    // exact single greedy step.
    const adaptive = this.specDecodeAdaptive;
    const breakEven = this.specDecodeCostRatio - 1; // E[a] threshold
    const minMatch = this.specDecodeMinMatch;
    // EWMA of accepted-tokens-per-drafted-round (a). Seed OPTIMISTIC (at the cap)
    // so we probe early; the first few rounds calibrate it to the real workload.
    const EWMA_ALPHA = 0.3;
    let ewmaAccept = kCap; // optimistic prior → draft until proven unprofitable
    let kCur = kCap; // current adaptive draft length
    // Cooldown: when we back off, skip drafting for a few rounds, then probe.
    let cooldown = 0;
    const COOLDOWN_ROUNDS = 8; // rounds to coast on cheap single-token steps
    const PROBE_K = 2; // small draft used to re-test the waters cheaply

    // Emit one token; returns false if generation must stop (stop token or
    // budget reached) — mirrors the baseline loop's per-token bookkeeping.
    const emit = (tokenId: number): boolean => {
      if (outLen - promptLen >= maxNew) return false;
      if (this.stopTokens.has(tokenId)) return false;
      generatedTokens.push(tokenId);
      args.onToken?.({
        tokenId,
        tokenIndex: generatedTokens.length - 1,
        maxNewTokens: maxNew,
      });
      terminalTailState = this.updateTerminalTailState(tokenId, terminalTailState);
      out[outLen++] = tokenId;
      seq.push(tokenId);
      return true;
    };

    // ── Seed the first token (lossless). ──────────────────────────────────
    // The host owns the prompt's last-token distribution (`args.prefill.logits`)
    // — the prefill path does NOT leave it on the station's `self.logits`. We
    // pick the first token here (pure argmax; spec is gated on the default
    // sampler with no active prosody bias, so `chooseToken` == argmax, identical
    // to the baseline loop) and COMMIT it via `/decode-next`. That forward writes
    // the token's KV at slot `promptLen`, advances `self.pos` to `promptLen + 1`,
    // and — crucially — leaves `self.logits` holding the genuine model
    // distribution for the next slot, so every later verify-batch round's i==0 /
    // empty-candidate path reads a CORRECT distribution.
    const firstToken = this.chooseToken(args.prefill.logits, {
      emittedTokens: 1,
      maxNewTokens: maxNew,
      generatedTokens,
      tailState: terminalTailState,
      stopTokens: this.stopTokens,
    });
    if (this.stopTokens.has(firstToken)) return finish();
    try {
      // Return value (the model's argmax for slot promptLen+1) is discarded: the
      // first verify-batch round recomputes it from the now-valid self.logits.
      await decodeNext.call(this.entryStation, {
        tokenId: firstToken,
        position: promptLen,
        requestId: this.requestId,
      });
    } catch (error) {
      // No emission yet → fall back to the baseline loop on untouched state.
      if (this.decodeNextUnavailable(error) && outLen === startOutLen) return null;
      throw error;
    }
    if (!emit(firstToken)) return finish();

    while (outLen - promptLen < maxNew) {
      const budget = maxNew - (outLen - promptLen); // tokens still wanted
      const maxDraft = Math.max(0, budget - 1); // verify always also emits nextToken

      // ── Decide this round's draft. ────────────────────────────────────────
      // Non-adaptive (legacy) path: draft on any match at the fixed cap.
      // Adaptive path: gate on match strength + the rolling cost-gate, with a
      // cooldown/probe rhythm and an adaptive K. Either way `draft` may be EMPTY,
      // which degrades to exactly one cheap single-token greedy step (lossless).
      let draft: number[];
      if (!adaptive) {
        draft = ngramDraft(seq, { k: kCap }).slice(0, maxDraft);
      } else if (cooldown > 0) {
        // Backed off: coast on cheap single-token steps. Drop to a 1-token round
        // unless this is the final probe round, where we draft a SMALL K to
        // re-test acceptance cheaply (so we recover when text turns repetitive).
        cooldown -= 1;
        if (cooldown === 0) {
          const { draft: probe, matchLen } = ngramDraftWithMatch(seq, {
            k: Math.min(PROBE_K, kCap),
          });
          // Only probe if there is at least a strong-ish match to test.
          draft =
            matchLen >= minMatch ? probe.slice(0, maxDraft) : [];
        } else {
          draft = []; // cheap single-token step
        }
      } else {
        // Active: draft only when (a) the match is long enough (strong signal)
        // and (b) the EWMA acceptance clears the break-even E[a] > R-1. Below
        // break-even → BACK OFF: empty draft + start a cooldown.
        const { draft: cand, matchLen } = ngramDraftWithMatch(seq, { k: kCur });
        if (matchLen >= minMatch && ewmaAccept > breakEven) {
          draft = cand.slice(0, maxDraft);
        } else {
          draft = [];
          if (ewmaAccept <= breakEven) cooldown = COOLDOWN_ROUNDS;
        }
      }

      let result: Awaited<ReturnType<NonNullable<IStation['verifyBatch']>>>;
      try {
        result = await verifyBatch.call(this.entryStation, {
          candidates: draft,
          // ABSOLUTE KV slot of the first candidate == the running seq length.
          // Drives `self.pos` on the fat-station so the candidate forward writes
          // KV at the right slots and never clobbers the warm prompt KV. After
          // the seed `seq.length == self.pos`, so this stays in lockstep.
          position: seq.length,
          requestId: this.requestId,
        });
      } catch (error) {
        if (this.verifyBatchUnavailable(error) && outLen === startOutLen) {
          // Station can't do /verify-batch and we've emitted nothing yet → let
          // the caller run the baseline loop on the untouched station state.
          return null;
        }
        throw error;
      }

      // ── Update the cost-gate state from this round's OBSERVED acceptance. ──
      // Only drafted rounds (candidateCount > 0) carry acceptance signal; a
      // single-token round is the back-off itself and would dilute the estimate
      // toward 0, so we leave the EWMA untouched on empty drafts. accepted-per-
      // drafted-round (a) is what the break-even E[a] > R-1 is defined over.
      if (adaptive && result.candidateCount > 0) {
        const a = result.acceptedCount;
        ewmaAccept = EWMA_ALPHA * a + (1 - EWMA_ALPHA) * ewmaAccept;
        // Adaptive K: shrink toward 1 when this draft was mostly rejected (less
        // wasted forward + cheaper wrong drafts), grow back toward the cap when
        // it accepted well. Heuristic: full accept → +1; zero accept → halve.
        if (a >= result.candidateCount) {
          kCur = Math.min(kCap, kCur + 1);
        } else if (a === 0) {
          kCur = Math.max(1, kCur >> 1);
        }
        // else: partial accept → keep K (it's roughly right-sized).
      }

      // Emit accepted draft prefix ++ [nextToken]; stop at budget / stop token.
      let halted = false;
      for (const t of result.accepted) {
        if (!emit(t)) {
          halted = true;
          break;
        }
      }
      if (!halted && !emit(result.nextToken)) halted = true;
      if (halted) break;
    }

    return finish();
  }

  async generateFromPrefill(
    args: PipelineGenerateFromPrefillArgs,
  ): Promise<Uint32Array> {
    if (!args.promptTokens.length) {
      throw new Error('Pipeline.generateFromPrefill: promptTokens must be non-empty');
    }
    if (!Number.isInteger(args.maxNewTokens) || args.maxNewTokens < 0) {
      throw new Error(
        `Pipeline.generateFromPrefill: maxNewTokens must be a non-negative integer (got ${String(args.maxNewTokens)})`,
      );
    }

    const out = new Uint32Array(args.promptTokens.length + args.maxNewTokens);
    out.set(args.promptTokens);
    let outLen = args.promptTokens.length;
    const finish = (): Uint32Array =>
      outLen === out.length ? out : out.slice(0, outLen);
    if (args.prefill.poisoned || args.maxNewTokens === 0) {
      return finish();
    }

    // ── Prompt-lookup / n-gram speculative decode (DEFAULT OFF). ───────────
    // Greedy + LOSSLESS: every emitted token is an argmax, so with the same
    // prompt + temp=0 the output is byte-identical to the non-speculative loop
    // below. Eligible only for the greedy single-station argmax path (no
    // sampler / steering / terminal-prosody / alphaJump — those are not
    // greedy-argmax). When ineligible (or the flag is off) we fall straight
    // through to the unchanged loop.
    if (this.shouldSpeculate(args.alphaJump !== undefined)) {
      const spec = await this.generateFromPrefillSpeculative(args, out, outLen);
      if (spec !== null) return spec;
      // null → the station rejected /verify-batch (404 / not-available); fall
      // through to the baseline loop with the station state untouched.
    }

    // Phase 2: autoregressive single-token generate. Each token runs through
    // every station once. chunkSize is implicitly 1; we reuse the same
    // dispatch primitives for clarity, but a single chunk per token does not
    // need Promise.race over multiple lanes.
    const generatedTokens: number[] = [];
    let terminalTailState =
      this.terminalProsody?.tailState ?? 'open-clause';
    let nextToken = this.chooseToken(args.prefill.logits, {
      emittedTokens: 1,
      maxNewTokens: args.maxNewTokens,
      generatedTokens,
      tailState: terminalTailState,
      stopTokens: this.stopTokens,
    });
    if (this.stopTokens.has(nextToken)) {
      return finish();
    }
    generatedTokens.push(nextToken);
    args.onToken?.({
      tokenId: nextToken,
      tokenIndex: generatedTokens.length - 1,
      maxNewTokens: args.maxNewTokens,
    });
    terminalTailState = this.updateTerminalTailState(
      nextToken,
      terminalTailState,
    );
    out[outLen++] = nextToken;

    for (let step = 1; step < args.maxNewTokens; step++) {
      const pos = outLen - 1;
      const samplingContext: PipelineSamplingContext = {
        emittedTokens: step + 1,
        maxNewTokens: args.maxNewTokens,
        generatedTokens,
        tailState: terminalTailState,
        stopTokens: this.stopTokens,
      };
      const decodeNext = this.entryStation.decodeNext;
      if (decodeNext && this.canUseRemoteDecode(samplingContext)) {
        try {
          const decodeArgs: Parameters<NonNullable<typeof decodeNext>>[0] = {
            tokenId: nextToken,
            position: pos,
            requestId: this.requestId,
          };
          // Native terminal-prosody: when cadence pressure has crossed the
          // Reynolds break, hand the prosody params to `/decode-next` so the
          // bias is applied natively on the forward's own logits — no slow
          // logits-pull fallback, no extra HTTP hops.
          const prosody = this.remoteDecodeProsody(samplingContext);
          if (prosody) decodeArgs.terminalProsody = prosody;
          nextToken = await decodeNext.call(this.entryStation, decodeArgs);
          this.emitObservedTerminalProsody(
            samplingContext,
            nextToken,
            'gnosis:pipeline:decode-next:observe',
          );
        } catch (error) {
          if (!this.decodeNextUnavailable(error)) throw error;
          const fallbackToken = await this.chooseTokenThroughStations(
            nextToken,
            pos,
            samplingContext,
            args.alphaJump,
            args.prefixHash,
          );
          if (fallbackToken === null) return finish();
          nextToken = fallbackToken;
        }
      } else {
        const fallbackToken = await this.chooseTokenThroughStations(
          nextToken,
          pos,
          samplingContext,
          args.alphaJump,
          args.prefixHash,
        );
        if (fallbackToken === null) return finish();
        nextToken = fallbackToken;
      }
      if (this.stopTokens.has(nextToken)) {
        break;
      }
      generatedTokens.push(nextToken);
      args.onToken?.({
        tokenId: nextToken,
        tokenIndex: generatedTokens.length - 1,
        maxNewTokens: args.maxNewTokens,
      });
      terminalTailState = this.updateTerminalTailState(
        nextToken,
        terminalTailState,
      );
      out[outLen++] = nextToken;
    }

    return finish();
  }

  /**
   * Just the prefill phase. Useful for benchmarks and prefix caching.
   *
   * Implementation mirrors `pipelinedPrefill` from gcp-coordinator.ts:
   *   - Build chunks of size `floor(P/N)` (or explicit override)
   *   - Seed the pipeline by dispatching chunk 0 to station 0
   *   - Loop: race in-flight, advance completed chunk to next station OR
   *     retire it. Whenever a station finishes its last stage, free station 0
   *     for the next unstarted chunk.
   */
  async pipelinedPrefill(args: {
    promptTokens: Uint32Array;
    startPosOffset?: number;
    computeLogits?: boolean;
    captureStationResiduals?: boolean;
    onProgress?: (progress: PipelinePrefillProgress) => void;
  }): Promise<PrefillResult> {
    const numTokens = args.promptTokens.length;
    if (!numTokens) {
      throw new Error('Pipeline.pipelinedPrefill: promptTokens must be non-empty');
    }
    const startPosOffset = args.startPosOffset ?? 0;
    if (!Number.isInteger(startPosOffset) || startPosOffset < 0) {
      throw new Error(
        `Pipeline.pipelinedPrefill: startPosOffset must be a non-negative integer (got ${String(startPosOffset)})`,
      );
    }

    const numNodes = this.stations.length;
    const lastNode = numNodes - 1;
    const hiddenDim = this.hiddenDim;

    // TQFT Cobordism: when cobordism stations are configured, replace
    // the single entryStation.embed() with parallel fan-out to all
    // cobordism shards and merge via the Frobenius identity (μ: A ⊗ A → A).
    // Since each shard covers a disjoint vocab range, exactly one shard
    // produces non-zero output per token, making the merge equivalent
    // to elementwise addition.
    let allEmbeddings: Float32Array;
    if (this.cobordismStations && this.cobordismStations.length > 0) {
      // Fan-out to all cobordism shards in parallel
      const partialResults = await Promise.all(
        this.cobordismStations.map((cs) =>
          cs.station.cobordismEmbed({
            tokens: args.promptTokens,
            startPos: startPosOffset,
            vocabStart: cs.vocabStart,
            vocabEnd: cs.vocabEnd,
            requestId: this.requestId,
          }),
        ),
      );

      // Frobenius merge: elementwise addition of disjoint partial embeddings
      const totalFloats = numTokens * hiddenDim;
      allEmbeddings = new Float32Array(totalFloats);
      for (const partial of partialResults) {
        if (partial.length !== totalFloats) {
          throw new Error(
            `Pipeline cobordism merge: shard returned ${partial.length} floats, expected ${totalFloats}`,
          );
        }
        for (let j = 0; j < totalFloats; j++) {
          allEmbeddings[j]! += partial[j]!;
        }
      }
    } else {
      // Legacy path: single station /embed call
      allEmbeddings = await this.entryStation.embed({
        tokens: args.promptTokens,
        startPos: startPosOffset,
        requestId: this.requestId,
      });
    }
    let stationCallCount = 1; // the embed call(s)

    const expected = numTokens * hiddenDim;
    if (allEmbeddings.length !== expected) {
      throw new Error(
        `Pipeline.pipelinedPrefill: embed returned ${allEmbeddings.length} floats, expected ${expected} (numTokens=${numTokens} * hiddenDim=${hiddenDim})`,
      );
    }

    // Chunk size: explicit override or floor(P / N), clamped to [1, BATCH_CAP].
    //
    // The Rust kernel caps batched-token state at `BATCH_CAPACITY` per call:
    //   wasm32 target:  4   (apps/distributed-inference-worker, *.workers.dev)
    //   native target:  8   (fat-station local mesh, bench harnesses)
    // (see open-source/gnosis/distributed-inference/src/model.rs:381-383).
    //
    // Without this clamp, the default heuristic (floor(numTokens / numNodes))
    // produces chunkSize=15 for a 30-token chat-templated prompt across 2
    // stations — which exceeds both caps and panics the kernel with
    // `seq_len 15 out of range (0, 8]`. On wasm the panic gets re-wrapped by
    // wasm-bindgen into the cryptic "recursive use of an object detected
    // which would lead to unsafe aliasing in rust" error, which is the
    // root cause of the deployed-mesh /forward 500s observed 2026-04-26.
    //
    // We default to 4 (the wasm safe cap) because the Pipeline does not
    // know which target backs each Station behind HTTP — assuming the
    // tighter of the two caps is correct on both. Callers with a known-
    // native mesh can raise it via `chunkSize: 8` in the config.
    // Honor explicit overrides verbatim (caller knows their target);
    // otherwise default to floor(P/N) clamped to the per-station cap.
    //
    // Each Station declares `maxBatchSize` (8 on native HTTP/UDS/UDS-flow/
    // UDP-flow, 4 on wasm-backed Workers). The pipeline takes the MIN over
    // all stations: any chunk traverses every station, so one wasm-bounded
    // station forces the whole chain to its cap. Stations that omit
    // `maxBatchSize` fall back to the wasm-safe floor of 4 (backward
    // compatibility for pre-2026-04-26 IStation implementations).
    const BATCH_CAP_SAFE = Math.min(...this.stations.map((s) => s.maxBatchSize ?? 4));
    const chunkSize = Math.max(
      1,
      this.explicitChunkSize ??
        Math.min(BATCH_CAP_SAFE, Math.floor(numTokens / Math.max(numNodes, 1))),
    );

    const chunks: ChunkDescriptor[] = [];
    for (let pos = 0; pos < numTokens; pos += chunkSize) {
      const count = Math.min(chunkSize, numTokens - pos);
      chunks.push({ startPos: pos, count });
    }
    const numChunks = chunks.length;

    const getChunkEmbeddings = (chunk: ChunkDescriptor): Float32Array => {
      const offset = chunk.startPos * hiddenDim;
      return allEmbeddings.subarray(offset, offset + chunk.count * hiddenDim);
    };

    // Per-node in-flight slot. Mirrors aether coordinator naming verbatim.
    const inFlight: (Promise<InFlightResult> | null)[] = new Array(numNodes).fill(null);

    const dispatch = (nodeIdx: number, chunkIdx: number, input: Float32Array): void => {
      const chunk = chunks[chunkIdx]!;
      const absoluteStartPos = startPosOffset + chunk.startPos;
      const station = this.stations[nodeIdx]!;
      stationCallCount++;
      const copied = copyView(input);
      const spectralNoiseMonitor = this.spectralNoiseMonitor;
      const dispatchedAt = spectralNoiseMonitor ? Date.now() : 0;
      if (TRACE) {
        traceLog(
          `DISPATCH chunk=${chunkIdx} node=${nodeIdx} startPos=${absoluteStartPos} ` +
            `count=${chunk.count} residual_len=${copied.length}${traceFirst4(copied, 'in')}`,
        );
      }
      inFlight[nodeIdx] = (async (): Promise<InFlightResult> => {
        // FANO CACHE (hot path): station.forward is PURE in (residual, startPos,
        // seqLen, station). A content-key hit is a geodesicLength:0 replay that
        // skips the (often bound-service) forward entirely — FREE, zero burn.
        let key: string | null = null;
        if (this.enableChunkCache) {
          key = await this.chunkKey(copied, nodeIdx, absoluteStartPos, chunk.count);
          const cached = this.chunkCache.get(key);
          if (cached) {
            this.chunkCacheHits++;
            return { nodeIdx, chunkIdx, result: copyView(cached) };
          }
          this.chunkCacheMisses++;
        }
        const reply = await station.forward({
          residual: copied,
          startPos: absoluteStartPos,
          sequenceLength: chunk.count,
          requestId: this.requestId,
        });
        this.chunkTokenBurn += chunk.count; // forwards burn the real work; hits are free
        if (spectralNoiseMonitor) {
          const completedAt = Date.now();
          spectralNoiseMonitor.ingest({
            requestId: this.requestId,
            nodeIdx,
            chunkIdx,
            startPos: absoluteStartPos,
            sequenceLength: chunk.count,
            latencyMs: Math.max(0, completedAt - dispatchedAt),
            requestEntropy: requestEntropy(chunk.count, nodeIdx),
            timestamp: completedAt,
          });
        }
        if (TRACE) {
          traceLog(
            `COMPLETE chunk=${chunkIdx} node=${nodeIdx} result_len=${reply.residual.length}` +
              traceFirst4(reply.residual, 'out'),
          );
        }
        if (this.enableTeleportation && this.teleportationOptimizer && reply.teleportation) {
          this.updateTeleportation(reply.teleportation);
        }
        // Cache a SNAPSHOT (copy) so downstream mutation of the result can't
        // corrupt the cached value.
        if (key !== null) this.chunkCachePut(key, copyView(reply.residual));
        return { nodeIdx, chunkIdx, result: reply.residual };
      })();
    };

    // ── Bizarro companion state (flag-gated, off by default) ──────────────
    //
    // When enabled, a bizarro companion speculatively computes a reject-mask
    // forward for the next unstarted chunk whenever a downstream pipeline
    // slot is idle. Retirement uses the rule:
    //   argmax(nominal_logits) ∉ rejectMask  →  count as bizarro retirement
    //
    // Mirrors `gcp-coordinator.ts` :: bizarroMasks / dispatchBizarroCompanion
    // / countIdleDownstreamSlots. Backed by BizarroCheckpointHomology.lean
    // and AeonFlowBizarro.lean.
    //
    // Critical: speculation is fire-and-forget. Errors from `bizarroForward`
    // are already swallowed by the Station; here we additionally guard the
    // promise chain so an unexpected throw cannot propagate to the main
    // dispatch loop. The returned promise is intentionally not awaited and
    // never enters the `Promise.race` pool — only the `.then` mask-stash
    // observes its outcome.
    const bizarroMasks = new Map<number, number[]>();
    let bizarroDispatches = 0;
    let bizarroRetirements = 0;
    const enableBizarro = this.enableBizarro;
    const entryStation = this.entryStation;

    const countIdleDownstreamSlots = (): number => {
      let n = 0;
      for (let i = 1; i < numNodes; i++) {
        if (inFlight[i] === null) n++;
      }
      return n;
    };

    const dispatchBizarroCompanion = (
      targetChunkIdx: number,
      input: Float32Array,
    ): void => {
      if (!enableBizarro) return;
      const chunk = chunks[targetChunkIdx]!;
      bizarroDispatches++;
      // Copy view so the speculative path doesn't share buffer state with
      // the nominal dispatch (matches `dispatch`'s `copyView(input)`).
      const residual = copyView(input);
      // Fire-and-forget: capture the promise inside its own try so neither
      // the synchronous call nor the .then/.catch can leak. The orphan
      // promise stashes the mask on success and silently drops on failure.
      try {
        void entryStation
          .bizarroForward({
            residual,
            startPos: startPosOffset + chunk.startPos,
            sequenceLength: chunk.count,
            targetPosition: startPosOffset + chunk.startPos + chunk.count - 1,
            requestId: this.requestId,
          })
          .then((r: { rejectMask: number[] } | null) => {
            if (r && r.rejectMask.length) {
              bizarroMasks.set(targetChunkIdx, r.rejectMask);
            }
          })
          .catch(() => {
            // Speculative — drop silently.
          });
      } catch {
        // Synchronous throw from bizarroForward (shouldn't happen, but the
        // contract is "never poison nominal"). Drop silently.
      }
    };

    let nextChunkToDispatch = 0;
    let completedChunks = 0;
    let completedTokenCount = 0;
    const emitProgress = (): void => {
      args.onProgress?.({
        completedTokens: completedTokenCount,
        totalTokens: numTokens,
        completedChunks,
        totalChunks: numChunks,
      });
    };
    // Typed as `Float32Array<ArrayBufferLike>` so we can hold either the
    // initial zero-buffer or a value coming back from a Station fetch
    // response (which TS narrows as ArrayBufferLike under strict mode).
    let lastHiddenStates: Float32Array<ArrayBufferLike> = new Float32Array(
      hiddenDim * (chunks[numChunks - 1]?.count ?? 1),
    );
    let poisonChunk = -1;

    // Pending-advance queue. When a chunk exits node K but node K+1 is still
    // busy, we stash its residual here so the next free-slot scan can place
    // it. Required for adversarial latency patterns (fast upstream, slow
    // downstream) where multiple chunks may queue ahead of a slow station.
    // FIFO per destination node guarantees in-order delivery.
    const advancePending: { chunkIdx: number; toNodeIdx: number; input: Float32Array }[] = [];
    const completedByStation: Array<Array<Float32Array | undefined>> | undefined =
      args.captureStationResiduals === true
        ? Array.from(
            { length: numNodes },
            () => new Array<Float32Array | undefined>(numChunks),
          )
        : undefined;

    const tryFlushAdvancePending = (): void => {
      // Single linear scan: dispatch the head item for every destination
      // whose slot is now free. Items are kept in their relative order so
      // chunks land on a station in the order they exited the prior one.
      const remaining: typeof advancePending = [];
      const claimed = new Set<number>();
      for (const p of advancePending) {
        if (!inFlight[p.toNodeIdx] && !claimed.has(p.toNodeIdx)) {
          dispatch(p.toNodeIdx, p.chunkIdx, p.input);
          claimed.add(p.toNodeIdx);
        } else {
          remaining.push(p);
        }
      }
      advancePending.length = 0;
      advancePending.push(...remaining);
    };

    // Seed: chunk 0 -> station 0.
    emitProgress();
    dispatch(0, 0, getChunkEmbeddings(chunks[0]!));
    nextChunkToDispatch = 1;

    while (completedChunks < numChunks) {
      const active: Promise<InFlightResult>[] = [];
      for (let i = 0; i < numNodes; i++) {
        if (inFlight[i]) active.push(inFlight[i]!);
      }
      if (!active.length) break;

      const completed = await Promise.race(active);
      inFlight[completed.nodeIdx] = null;

      const chunkTokenCount = chunks[completed.chunkIdx]!.count;
      const lastTokenOffset = (chunkTokenCount - 1) * hiddenDim;
      const lastTokenStates = completed.result.subarray(
        lastTokenOffset,
        lastTokenOffset + hiddenDim,
      );
      if (completedByStation) {
        completedByStation[completed.nodeIdx]![completed.chunkIdx] = copyView(completed.result);
      }

      if (hasNaN(lastTokenStates) && poisonChunk < 0) {
        poisonChunk = completed.chunkIdx;
      }

      const isPoisoned = poisonChunk >= 0 && completed.chunkIdx >= poisonChunk;

      if (completed.nodeIdx < lastNode && !isPoisoned) {
        // Forward this chunk's residual to the next station -- but only if
        // that station is free AND no earlier-queued chunk is waiting for
        // the same slot. Otherwise enqueue.
        const toNodeIdx = completed.nodeIdx + 1;
        const slotQueued = advancePending.some((p) => p.toNodeIdx === toNodeIdx);
        
        // Teleportation skip: if all layers in the NEXT station's range are bypassed,
        // treat it as completed and hop further downstream immediately.
        if (this.enableTeleportation && this.teleportationOptimizer) {
          let nextIdx = toNodeIdx;
          while (nextIdx < numNodes) {
             const bypassed = this.getBypassedStations();
             if (bypassed.has(nextIdx)) {
               traceLog(`TELEPORT SKIP chunk=${completed.chunkIdx} node=${nextIdx} range=${this.stations[nextIdx]!.startLayer}-${this.stations[nextIdx]!.endLayer}`);
               nextIdx++;
             } else {
               break;
             }
          }
          if (nextIdx >= numNodes) {
            // All downstream nodes bypassed! Retire now.
            lastHiddenStates = completed.result;
            completedChunks++;
            completedTokenCount += chunks[completed.chunkIdx]!.count;
            emitProgress();
            tryFlushAdvancePending();
            continue;
          }
          // Hop to the first non-bypassed node
          const targetNodeIdx = nextIdx;
          if (!inFlight[targetNodeIdx] && !advancePending.some(p => p.toNodeIdx === targetNodeIdx)) {
             dispatch(targetNodeIdx, completed.chunkIdx, completed.result);
          } else {
             advancePending.push({
               chunkIdx: completed.chunkIdx,
               toNodeIdx: targetNodeIdx,
               input: completed.result,
             });
          }
        } else {
          if (!inFlight[toNodeIdx] && !slotQueued) {
            dispatch(toNodeIdx, completed.chunkIdx, completed.result);
          } else {
            advancePending.push({
              chunkIdx: completed.chunkIdx,
              toNodeIdx,
              input: completed.result,
            });
          }
        }
      } else {
        // Either the chunk exited the last station or it's been poisoned.
        //
        // Bizarro retirement hook: if a companion mask is ready for this
        // chunk AND the chunk isn't poisoned AND its nominal hidden states
        // are clean, count it as an early retirement. The nominal logits
        // remain authoritative — the manuscript's "skip recheck" semantics
        // applies to multi-token verification we don't have yet.
        if (!isPoisoned) {
          lastHiddenStates = completed.result;
          if (enableBizarro) {
            const mask = bizarroMasks.get(completed.chunkIdx);
            if (mask && !hasNaN(lastTokenStates)) {
              // We don't have logits-per-chunk yet (lm_head only fires once
              // at the end of prefill), so we substitute argmax over the
              // last-token hidden state vs the mask. This is consistent
              // with aether's reference: `bizarroRetirements++` is purely
              // a counter; retirement is observability, not an algorithmic
              // shortcut, until lm_head per-chunk lands.
              const pick = argmax(lastTokenStates);
              if (!mask.includes(pick)) {
                bizarroRetirements++;
              }
            }
          }
        }
        completedChunks++;
        completedTokenCount += chunks[completed.chunkIdx]!.count;
        emitProgress();
      }

      // Wallington Rotation: when station 0 frees up (either because a chunk
      // exited the last node, or because the chunk we held just hopped
      // downstream), accept the next unstarted chunk.
      //
      // BUG FIX 2026-04-27: previously we dispatched to node 0 whenever
      // the LAST node finished. But node 0 might still be busy with a
      // different chunk! We must strictly check !inFlight[0] before
      // overwriting the slot with a new chunk.
      if (
        nextChunkToDispatch < numChunks &&
        poisonChunk < 0 &&
        !inFlight[0]
      ) {
        const idx = nextChunkToDispatch++;
        const targetEmbeddings = getChunkEmbeddings(chunks[idx]!);
        dispatch(0, idx, targetEmbeddings);
        if (enableBizarro && poisonChunk < 0 && countIdleDownstreamSlots() > 0) {
          dispatchBizarroCompanion(idx, targetEmbeddings);
        }
      }

      // Try to flush any pending advances only while the pipeline is still
      // healthy. Once poison is observed, queued downstream dispatches are
      // no longer allowed to start; only already in-flight work may settle.
      if (poisonChunk < 0) {
        tryFlushAdvancePending();
      } else {
        advancePending.length = 0;
      }

      // Drain on poison: stop dispatching, but let in-flight chunks ahead
      // settle so promises don't dangle.
      if (poisonChunk >= 0) {
        const stillActive = inFlight.some((p) => p !== null);
        if (!stillActive) break;
      }
    }

    // Apply final-norm + lm_head to the last token of the last chunk that
    // survived the pipeline. If everything was poisoned, return NaN-laced
    // logits via lmHead -- the caller decides what to do.
    const lastChunkCount = chunks[numChunks - 1]!.count;
    const tailResidual =
      lastHiddenStates.length === hiddenDim
        ? lastHiddenStates
        : lastHiddenStates.subarray(
            (lastChunkCount - 1) * hiddenDim,
            lastChunkCount * hiddenDim,
          );
    const stationPrefixResiduals = completedByStation?.map((chunksForStation) => {
      const total = chunksForStation.reduce(
        (sum, chunk) => sum + (chunk?.length ?? 0),
        0,
      );
      if (total !== numTokens * hiddenDim) return undefined;
      const out = new Float32Array(total);
      let offset = 0;
      for (const chunk of chunksForStation) {
        if (!chunk) return undefined;
        out.set(chunk, offset);
        offset += chunk.length;
      }
      return out;
    });

    // TQFT comultiplication (Δ: A → A ⊗ A): when cobordism shards are
    // configured, the exit station does NOT hold `output_weight` (~447 MiB
    // Q6_K — too big for a 128 MiB Worker). Instead, fan out lm_head across
    // the same vocab-sliced shards that own the embedding rows (qwen-coder
    // ties its embeddings, so the rows ARE the lm_head weights). Each shard
    // returns partial logits over its vocab range; we concatenate into the
    // full vocab vector. Without this, the exit station's /lm-head call
    // panics on `tensor 'output_weight' not pre-loaded`.
    let logits: Float32Array;
    if (args.computeLogits === false) {
      logits = new Float32Array(0);
    } else if (this.cobordismStations && this.cobordismStations.length > 0) {
      if (process.env['PIPELINE_TRACE']) {
        console.log(
          `[PIPE_TRACE] LM_HEAD COBORDISM_FANOUT shards=${this.cobordismStations.length}`,
        );
      }
      stationCallCount++;
      const partials = await Promise.all(
        this.cobordismStations.map((cs, i) =>
          cs.station.cobordismLmHead({
            residual: tailResidual,
            sequenceLength: 1,
            vocabStart: cs.vocabStart,
            vocabEnd: cs.vocabEnd,
            requestId: this.requestId,
          }).then((res: Float32Array) => {
             if (process.env['PIPELINE_TRACE']) {
               console.log(`[PIPE_TRACE] LM_HEAD SHARD_COMPLETE shard=${i} len=${res.length}`);
             }
             return res;
          }),
        ),
      );
      if (process.env['PIPELINE_TRACE']) {
        console.log(`[PIPE_TRACE] LM_HEAD COBORDISM_MERGE`);
      }
      // Concatenate partials in vocab order. The cobordism config carries
      // [vocabStart, vocabEnd) per shard; the shards must cover the full
      // vocab without gaps (verified at config time).
      const totalVocab = this.cobordismStations.reduce(
        (acc, cs) => Math.max(acc, cs.vocabEnd),
        0,
      );
      logits = new Float32Array(totalVocab);
      for (let i = 0; i < this.cobordismStations.length; i++) {
        const cs = this.cobordismStations[i]!;
        const partial = partials[i]!;
        if (partial.length !== cs.vocabEnd - cs.vocabStart) {
          throw new Error(
            `Pipeline cobordism lm_head: shard ${i} returned ${partial.length} floats, ` +
              `expected ${cs.vocabEnd - cs.vocabStart} for vocab [${cs.vocabStart}, ${cs.vocabEnd})`,
          );
        }
        logits.set(partial, cs.vocabStart);
      }
    } else {
      if (process.env['PIPELINE_TRACE']) {
        console.log(`[PIPE_TRACE] LM_HEAD EXIT_STATION_ONLY`);
      }
      logits = await this.exitStation.lmHead({
        residual: tailResidual,
        sequenceLength: 1,
        requestId: this.requestId,
      });
      stationCallCount++;
    }

    const completeStationPrefixResiduals =
      stationPrefixResiduals?.every((residual) => residual !== undefined) === true
        ? (stationPrefixResiduals as Float32Array[])
        : undefined;
    const result: PrefillResult = {
      logits,
      stationCallCount,
      poisoned: poisonChunk >= 0,
      poisonChunk,
      tailResidual: copyView(tailResidual),
      stationPrefixResiduals: completeStationPrefixResiduals,
    };
    if (enableBizarro) {
      result.bizarroDispatches = bizarroDispatches;
      result.bizarroRetirements = bizarroRetirements;
    }
    return result;
  }

  /**
   * Content key for a station's PURE forward — SHA-256 of the residual bytes
   * folded with (nodeIdx, startPos, seqLen). Collision-resistant ⇒ miss-not-lie.
   * FNV-1a fallback on runtimes without WebCrypto (honest: diverges → cache miss).
   */
  private async chunkKey(
    residual: Float32Array,
    nodeIdx: number,
    startPos: number,
    seqLen: number,
  ): Promise<string> {
    const scalars = `${nodeIdx}:${startPos}:${seqLen}:`;
    const c = (globalThis as typeof globalThis & { crypto?: Crypto }).crypto;
    if (!c?.subtle) {
      let h = 0xcbf29ce484222325n;
      const v = new Uint8Array(residual.buffer, residual.byteOffset, residual.byteLength);
      for (let i = 0; i < v.length; i++) {
        h ^= BigInt(v[i]!);
        h = (h * 0x100000001b3n) & 0xffffffffffffffffn;
      }
      return scalars + h.toString();
    }
    const bytes =
      residual.buffer instanceof ArrayBuffer
        ? new Uint8Array(residual.buffer, residual.byteOffset, residual.byteLength)
        : (() => {
            const cp = new Uint8Array(residual.byteLength);
            cp.set(new Uint8Array(residual.buffer, residual.byteOffset, residual.byteLength));
            return cp;
          })();
    const digest = await c.subtle.digest('SHA-256', bytes);
    const b = new Uint8Array(digest);
    let h = '';
    for (let i = 0; i < 12; i++) h += b[i]!.toString(16).padStart(2, '0');
    return scalars + h;
  }

  /** Insert into the per-chunk cache, evicting the oldest entry past the cap. */
  private chunkCachePut(key: string, residual: Float32Array): void {
    if (this.chunkCache.size >= this.chunkCacheCap) {
      const oldest = this.chunkCache.keys().next().value;
      if (oldest !== undefined) this.chunkCache.delete(oldest);
    }
    this.chunkCache.set(key, residual);
  }

  /**
   * Per-chunk fano cache telemetry. Hits teleport (geodesicLength:0, zero burn);
   * forwards burn `chunk.count` tokens (the real work) — meter to recoup, like the
   * sideways sidecar fan-out.
   */
  chunkCacheStats(): { hits: number; misses: number; tokenBurn: number; size: number } {
    return {
      hits: this.chunkCacheHits,
      misses: this.chunkCacheMisses,
      tokenBurn: this.chunkTokenBurn,
      size: this.chunkCache.size,
    };
  }

  private updateTeleportation(teleportation: { layerDeficits: number[]; rounds: number }): void {
    if (!this.teleportationOptimizer) return;
    
    // Construct a partial NeuralMeshTeleportationResult
    const result = {
      layerDeficits: teleportation.layerDeficits,
      numLayers: teleportation.layerDeficits.length,
      beta1: 9,
      compositeDimension: teleportation.layerDeficits.length * 9,
      rounds: teleportation.rounds,
      transmittedAt: Date.now(),
      layers: teleportation.layerDeficits.map((d, i) => ({
        layerIdx: i,
        // LayerConvergenceTrajectory requires `deficit`, `trajectory`,
        // `convergenceRound` in addition to the convergence flags. We only
        // have the current per-layer deficit on the wire — synthesize a
        // minimal trajectory `[d, d-1, ..., 0]` so the optimizer's downstream
        // math (dampening estimate, convergence round) has well-formed input.
        deficit: d,
        trajectory: Array.from({ length: d + 1 }, (_, k) => d - k),
        convergenceRound: d,
        isConverged: d === 0,
        dampeningEstimate: 0, // we don't have the full trajectory yet
      })),
      // totalDeficit is REQUIRED by NeuralMeshTeleportationResult — sum of
      // per-layer deficits straight from the wire payload.
      totalDeficit: teleportation.layerDeficits.reduce((acc, d) => acc + d, 0),
      meshHealth: 1.0,
      allConverged: teleportation.layerDeficits.every(d => d === 0),
    };

    const bypassedArr = this.teleportationOptimizer.getBypassedLayers(result);
    for (const layerIdx of bypassedArr) {
      this.bypassedLayers.add(layerIdx);
    }
  }

  private getBypassedStations(): Set<number> {
    const bypassed = new Set<number>();
    if (!this.enableTeleportation || !this.teleportationOptimizer) return bypassed;
    
    for (let i = 0; i < this.stations.length; i++) {
      const station = this.stations[i]!;
      let allBypassed = true;
      for (let layer = station.startLayer; layer < station.endLayer; layer++) {
        if (!this.bypassedLayers.has(layer)) {
          allBypassed = false;
          break;
        }
      }
      if (allBypassed) {
        bypassed.add(i);
      }
    }
    
    return bypassed;
  }
}

// ── helpers ────────────────────────────────────────────────────────────────

function hasNaN(arr: Float32Array): boolean {
  for (let i = 0; i < arr.length; i++) {
    if (Number.isNaN(arr[i]!)) return true;
  }
  return false;
}

function argmax(logits: Float32Array): number {
  let best = 0;
  let bestVal = -Infinity;
  for (let i = 0; i < logits.length; i++) {
    const v = logits[i]!;
    if (v > bestVal) {
      bestVal = v;
      best = i;
    }
  }
  return best;
}

/**
 * Copy a Float32Array view into a freshly-owned Float32Array. Necessary
 * because `getChunkEmbeddings` returns a subarray view of `allEmbeddings`,
 * and the Station serializes whatever buffer it's handed -- subarray slicing
 * preserves the parent buffer's full byteLength, which would over-send.
 */
function copyView(view: Float32Array): Float32Array {
  if (view.byteOffset === 0 && view.byteLength === view.buffer.byteLength) {
    return view;
  }
  const copy = new Float32Array(view.length);
  copy.set(view);
  return copy;
}
