// @a0n/.ignored_gnosis/distributed-inference-host/src/pleromatic-reasoning-validator.ts
/**
 * # Pleromatic Reasoning-Chain Validator — Observing-Mode Reasoning Validation
 *
 * Walks a multi-step reasoning trace and bails as soon as any step
 * fails to fold to the Pleromatic Grounding Triton (`closureChain 0
 * = 10`). Built on top of the `SovereignJudge` from `./pleromatic-core`.
 *
 * ## The early-exit / bail-fast design
 *
 * A "reasoning chain" is a sequence of steps where each step claims
 * a closure-tower position (e.g., `30` for depth-1 reasoning, `90`
 * for depth-2, `270` for depth-3). The validator walks the chain
 * step-by-step and short-circuits the moment a step's claimed
 * closure position is either (a) not in the closure tower or (b)
 * cannot be folded down to grounding. **No further steps are
 * processed.** This is a speed-first design: the validator's value
 * is rejecting bad chains as cheaply as possible.
 *
 * The streaming variant (`validateChainStreaming`) walks an
 * `Iterable<ReasoningStep>` with `for...of` + `break` on the first
 * failure, so a generator that hasn't yielded its tail will not be
 * forced to do so on a rejected chain. Lazy producers stay lazy.
 *
 * ## Design choices
 *
 * - **One `SovereignJudge` per validator instance.** The judge owns
 *   a `MonsterCache` whose closure-tower entries are reused across
 *   `validateChain` calls. Recreating the judge per call would
 *   re-warm the cache.
 * - **Plain-object verdicts.** Returning POJOs keeps the V8 hidden-
 *   class hot, avoids prototype walks, and lets the result be
 *   trivially `JSON.stringify`'d for cross-worker transport.
 * - **Closure index = fold-step count.** Per
 *   `Gnosis.PleromaticSovereignSieve.iterated_fold_descends_to_grounding`,
 *   `iteratedFold n (closureChain n) = 10`, so the number of fold
 *   steps required for a step at index `n` is exactly `n`. We sum
 *   indices to report `foldStepsTotal`.
 *
 * ## Lean source-of-truth
 *
 * @see open-source/gnosis-math/Gnosis/PleromaticSovereignSieve.lean
 * @see open-source/gnosis-math/Gnosis/PleromaticHigherClosures.lean
 * @see ./pleromatic-core.ts — `SovereignJudge`, `MonsterCache`, `FRF`
 */

import { PLEROMATIC_CLOSURE, SovereignJudge } from './pleromatic-core';
import { Verdict } from '@a0n/bitwise/triton-verdict';

/**
 * One step in a reasoning chain.
 *
 * `closurePosition` must be a closure-tower value
 * (`closureChain n = 10 × 3^n`): `10`, `30`, `90`, `270`, ... A
 * `label` and `metadata` are carried opaquely for trace reporting.
 *
 * `pending: true` marks a placeholder for a step that has NOT arrived
 * yet on a streaming source. It is not a reasoning fault — the chain
 * is simply still in flight. The streaming validator defers (`Abstain`)
 * on the first pending step rather than prematurely accepting the head
 * as a complete chain or rejecting it as broken.
 */
export interface ReasoningStep {
  closurePosition: number;
  label?: string;
  metadata?: Record<string, unknown>;
  pending?: boolean;
}

/**
 * Verdict returned by the chain validator.
 *
 * Carries the canonical 3-state TritonVerdict (`verdict`) alongside the
 * legacy `valid: boolean` discriminant, which is preserved for every
 * existing caller. The boolean could only say accept/reject and was
 * forced to read an unfinished, still-streaming chain as one or the
 * other; `verdict` restores the neutral middle:
 *
 *   - `valid: true`,  `verdict: Accept`   — chain fully validated.
 *   - `valid: false`, `verdict: Decline`  — a concrete reasoning fault
 *                                           (the named `reason`).
 *   - `valid: false`, `verdict: Abstain`, `deferred: true`
 *                                         — DEFERRED: the chain is
 *                                           incomplete (a `pending` step
 *                                           was reached before any fault).
 *                                           Don't reject a chain that is
 *                                           still arriving — fold/retry
 *                                           once the tail lands.
 *
 * Note the bool subsumption: a deferred (`Abstain`) chain reports
 * `valid: false` so no existing accept-path caller mistakes an
 * unfinished chain for a validated one. The third state is recoverable
 * only by inspecting `verdict` (or the `deferred` flag) — exactly the
 * information the bool drops. The deferred case is a member of the SAME
 * `valid: false` shape as a fault (its `reason` is `'deferred'`), so the
 * existing `if (!verdict.valid)` narrowing keeps working unchanged.
 */
export type ChainVerdict =
  | { valid: true; verdict: Verdict; stepCount: number; foldStepsTotal: number }
  | {
      valid: false;
      verdict: Verdict;
      failedAtIndex: number;
      failedStep: ReasoningStep;
      reason: 'not-in-tower' | 'below-grounding' | 'descent-failed' | 'deferred';
      /** Present (true) only when the verdict is a defer, not a fault. */
      deferred?: true;
      /** Index of the first `pending` step (defer case only). */
      pendingAtIndex?: number;
      /** Valid steps consumed before the defer (defer case only). */
      stepCount?: number;
      /** Fold-step budget of the consumed prefix (defer case only). */
      foldStepsTotal?: number;
    };

/**
 * The Pleromatic reasoning-chain validator.
 *
 * Single-instance hot path: hold one `SovereignJudge` (and its
 * `MonsterCache`) for the lifetime of the validator and reuse it
 * across every `validateChain` / `validateChainStreaming` call.
 *
 * Mirrors the Observing-mode validation primitive from
 * `Gnosis.PleromaticSovereignSieve` but lifted from a single
 * position to a finite chain of positions.
 */
export class ReasoningChainValidator {
  private readonly judge: SovereignJudge;

  constructor(judge: SovereignJudge = new SovereignJudge()) {
    this.judge = judge;
  }

  /**
   * Validate a fully-materialized chain. Walks the array and
   * short-circuits on the first failure.
   */
  validateChain(chain: readonly ReasoningStep[]): ChainVerdict {
    let foldStepsTotal = 0;
    for (let i = 0; i < chain.length; i++) {
      const step = chain[i];
      // A pending placeholder means the chain is still arriving — defer
      // (Abstain) rather than rejecting a chain that has not finished.
      if (step.pending === true) {
        return {
          valid: false,
          verdict: Verdict.Abstain,
          reason: 'deferred',
          failedAtIndex: i,
          failedStep: step,
          deferred: true,
          pendingAtIndex: i,
          stepCount: i,
          foldStepsTotal,
        };
      }
      const reason = this.validateOne(step);
      if (reason !== null) {
        return {
          valid: false,
          verdict: Verdict.Decline,
          failedAtIndex: i,
          failedStep: step,
          reason,
        };
      }
      // Closure index = fold-steps-to-grounding for this step.
      foldStepsTotal += this.judge.getCache().closureIndexOf(step.closurePosition);
    }
    return {
      valid: true,
      verdict: Verdict.Accept,
      stepCount: chain.length,
      foldStepsTotal,
    };
  }

  /**
   * Validate a lazy chain. Bails on the first failure *without*
   * consuming further steps from the iterator. A generator that
   * hasn't yielded its tail will not be forced past the failure.
   */
  validateChainStreaming(chainIter: Iterable<ReasoningStep>): ChainVerdict {
    let foldStepsTotal = 0;
    let stepCount = 0;
    let terminal: ChainVerdict | null = null;
    for (const step of chainIter) {
      // A pending placeholder means the producer has more steps coming but
      // they have not arrived yet — defer (Abstain). Like the failure path,
      // this breaks WITHOUT pulling the tail, so a lazy producer stays lazy.
      if (step.pending === true) {
        terminal = {
          valid: false,
          verdict: Verdict.Abstain,
          reason: 'deferred',
          failedAtIndex: stepCount,
          failedStep: step,
          deferred: true,
          pendingAtIndex: stepCount,
          stepCount,
          foldStepsTotal,
        };
        break;
      }
      const reason = this.validateOne(step);
      if (reason !== null) {
        terminal = {
          valid: false,
          verdict: Verdict.Decline,
          failedAtIndex: stepCount,
          failedStep: step,
          reason,
        };
        break;
      }
      foldStepsTotal += this.judge.getCache().closureIndexOf(step.closurePosition);
      stepCount++;
    }
    if (terminal !== null) return terminal;
    return { valid: true, verdict: Verdict.Accept, stepCount, foldStepsTotal };
  }

  /**
   * Single-step validation. Returns `null` if the step is valid;
   * otherwise the failure reason.
   */
  private validateOne(
    step: ReasoningStep,
  ): 'not-in-tower' | 'below-grounding' | 'descent-failed' | null {
    const k = step.closurePosition;
    if (k < PLEROMATIC_CLOSURE) return 'below-grounding';
    if (!this.judge.isValidPosition(k)) return 'not-in-tower';
    const v = this.judge.verifyDescent(k);
    if (!v.valid) return 'descent-failed';
    return null;
  }
}

/**
 * Coherence ratio: fraction of valid steps in a chain (0..1).
 *
 * Useful for partial-validation reporting where the chain is too
 * long or noisy to demand full validity but a quality signal is
 * still wanted. Walks the entire chain (no early exit) — call
 * `validateChain` instead when only the first failure matters.
 */
export function coherenceRatio(chain: readonly ReasoningStep[]): number {
  if (chain.length === 0) return 1;
  const judge = new SovereignJudge();
  let validCount = 0;
  for (const step of chain) {
    const k = step.closurePosition;
    if (k < PLEROMATIC_CLOSURE) continue;
    if (!judge.isValidPosition(k)) continue;
    if (judge.verifyDescent(k).valid) validCount++;
  }
  return validCount / chain.length;
}
