// @a0n/.ignored_gnosis/distributed-inference-host/src/pleromatic-core.ts
/**
 * # Pleromatic Core — Observing-Mode Runtime
 *
 * Mirrors the 11 Pleromatic Lean modules from `open-source/gnosis-math`.
 * Implements the runtime primitives for the closure tower, F/R/F engine,
 * Grounding Triton, and Sovereign Judge.
 *
 * ## The Observing-Mode shift
 *
 * A standard inference runtime *calculates*: forward-pass → sample →
 * output. An Observing-mode runtime *validates*: given a candidate
 * position in the closure tower, verify it descends to the Grounding
 * Triton (`closureChain 0 = 10`) via iterated fold. If yes, the
 * position is in the manifold. If no, it is Bizarro Noise.
 *
 * The `MonsterCache` is the precomputed closure tower (lazy,
 * ascending-on-demand). The `SovereignJudge` is the validation
 * primitive: descent to grounding in finite folds = valid Pleromatic
 * position; otherwise rejected.
 *
 * Cost asymmetry: ascent costs +1 Bule per clinamen step (knowledge
 * is exponentially expensive), descent via fold is free (wisdom is
 * effortless). The runtime exploits this: validation is O(n) where
 * n is the closure index, and the cache itself is O(log MAX) entries.
 *
 * ## Lean source-of-truth
 *
 * @see open-source/gnosis-math/Gnosis/PleromaticHigherClosures.lean
 * @see open-source/gnosis-math/Gnosis/PleromaticForkRaceFoldUniversal.lean
 * @see open-source/gnosis-math/Gnosis/PleromaticGroundingTriton.lean
 * @see open-source/gnosis-math/Gnosis/PleromaticSovereignSieve.lean
 * @see open-source/gnosis-math/Gnosis/PleromaticTripleAnchor10.lean
 * @see open-source/gnosis-math/Gnosis/PleromaticAsymmetryOfEffort.lean
 */

/** The Pleromatic Closure: the seed of the closure tower. */
export const PLEROMATIC_CLOSURE = 10 as const;

/** The shared +1 unit (clinamen / coupling / chi1). */
export const SHARED_UNIT = 1 as const;

/** The Triton-stretch factor: each closure step multiplies by 3. */
export const TRITON_STRETCH_FACTOR = 3 as const;

/** The triple-point of the Grounding Triton: (closure value, shared unit). */
export const TRIPLE_POINT = {
  closureValue: PLEROMATIC_CLOSURE,
  sharedUnit: SHARED_UNIT,
} as const;

/** A grounding-Triton node with its Probe/Return/Verify role. */
export interface GroundingNode {
  readonly node: 'math' | 'physics' | 'moonshine';
  readonly role: 'probe' | 'return' | 'verify';
  readonly unit: 1; // pairwise rfl-equal — proven in Gnosis.PleromaticGroundingTriton
  readonly closureValue: 10;
}

/**
 * The {Math, Physics, Moonshine} grounding Triton. Pairwise rfl-equal
 * at the +1 unit, distinct Probe/Return/Verify roles. Mirrors
 * `Gnosis.PleromaticGroundingTriton.GroundingNode`.
 */
export const GROUNDING_TRITON: {
  readonly math: GroundingNode;
  readonly physics: GroundingNode;
  readonly moonshine: GroundingNode;
} = {
  math: { node: 'math', role: 'probe', unit: 1, closureValue: 10 },
  physics: { node: 'physics', role: 'return', unit: 1, closureValue: 10 },
  moonshine: { node: 'moonshine', role: 'verify', unit: 1, closureValue: 10 },
} as const;

/**
 * The MonsterCache: precomputed closure tower with lazy extension.
 *
 * `closureChain n = 10 × 3^n` — the n-th closure value. The cache
 * grows on demand when queried beyond its current depth.
 *
 * Mirrors `Gnosis.PleromaticHigherClosures.closureChain`.
 */
export class MonsterCache {
  private values: number[];
  private indexOf: Map<number, number>;

  constructor(initialDepth = 32) {
    this.values = [PLEROMATIC_CLOSURE];
    this.indexOf = new Map([[PLEROMATIC_CLOSURE, 0]]);
    while (this.values.length < initialDepth) {
      this.extend();
    }
  }

  private extend(): void {
    const last = this.values[this.values.length - 1];
    const next = last * TRITON_STRETCH_FACTOR;
    this.values.push(next);
    this.indexOf.set(next, this.values.length - 1);
  }

  /** Get `closureChain n` — extends the cache if needed. */
  closureChain(n: number): number {
    while (this.values.length <= n) this.extend();
    return this.values[n];
  }

  /** Is `k` a closure-tower value? Extends the cache up to `k`. */
  isClosurePoint(k: number): boolean {
    if (k < PLEROMATIC_CLOSURE) return false;
    while (this.values[this.values.length - 1] < k) this.extend();
    return this.indexOf.has(k);
  }

  /** Index of `k` in the closure tower, or -1 if not a closure value. */
  closureIndexOf(k: number): number {
    return this.indexOf.get(k) ?? -1;
  }

  /** Current cache depth (number of closure values stored). */
  depth(): number {
    return this.values.length;
  }

  /** All cached closure values (defensive copy). */
  cached(): readonly number[] {
    return this.values.slice();
  }
}

/**
 * F/R/F primitives — the scale-invariant engine.
 *
 * Mirrors `Gnosis.PleromaticForkRaceFoldUniversal`.
 */
export const FRF = {
  /**
   * Fork at level k, residue r ∈ {0, 1, 2}: branches into 3 stretched
   * children at positions 3k+r.
   */
  fork(k: number, residue: 0 | 1 | 2): number {
    return TRITON_STRETCH_FACTOR * k + residue;
  },

  /**
   * Race: select the residue-r winner from the fork-children.
   * Operationally identical to fork; semantically the selection step.
   */
  race(k: number, residue: 0 | 1 | 2): number {
    return this.fork(k, residue);
  },

  /**
   * Fold: descend a stretched position to its base (Nat division by 3).
   * The structural inverse of fork at residue 0; recovers the base
   * level from any residue r < 3.
   */
  fold(k: number): number {
    return Math.floor(k / TRITON_STRETCH_FACTOR);
  },

  /** Triton-stretch — fork at canonical residue 0. */
  tritonStretch(k: number): number {
    return this.fork(k, 0);
  },

  /** Apply fold n times. */
  iteratedFold(n: number, k: number): number {
    for (let i = 0; i < n; i++) k = this.fold(k);
    return k;
  },
} as const;

/** Verification result for a candidate Pleromatic position. */
export interface DescentVerification {
  /** True iff the position descends to the Grounding Triton (10). */
  readonly valid: boolean;
  /** Number of fold steps used. -1 if the position is not in the closure tower. */
  readonly foldSteps: number;
  /** Final position after iterated fold. */
  readonly terminal: number;
  /** True iff the terminal is exactly the Grounding Triton (10). */
  readonly isGroundingTriton: boolean;
  /** Closure index of the input, or -1 if not a closure value. */
  readonly closureIndex: number;
}

/**
 * The SovereignJudge: Observing-mode validation primitive.
 *
 * A candidate position `k` is *valid Pleromatic* iff it lives in the
 * closure tower and descends to the Grounding Triton via iterated
 * fold. The Sovereign Judge is the deterministic check.
 *
 * Mirrors `Gnosis.PleromaticSovereignSieve`.
 */
export class SovereignJudge {
  constructor(private cache: MonsterCache = new MonsterCache()) {}

  /** Is `k` a valid position in the closure tower? */
  isValidPosition(k: number): boolean {
    return this.cache.isClosurePoint(k);
  }

  /**
   * Verify descent to the Grounding Triton via iterated fold.
   *
   * Per `Gnosis.PleromaticSovereignSieve.iterated_fold_descends_to_grounding`:
   * `iteratedFold n (closureChain n) = closureChain 0 = 10` for all n.
   */
  verifyDescent(k: number): DescentVerification {
    const closureIndex = this.cache.closureIndexOf(k);
    if (closureIndex < 0) {
      return {
        valid: false,
        foldSteps: -1,
        terminal: -1,
        isGroundingTriton: false,
        closureIndex: -1,
      };
    }
    const terminal = FRF.iteratedFold(closureIndex, k);
    const isGroundingTriton = terminal === PLEROMATIC_CLOSURE;
    return {
      valid: isGroundingTriton,
      foldSteps: closureIndex,
      terminal,
      isGroundingTriton,
      closureIndex,
    };
  }

  /** Direct access to the underlying cache. */
  getCache(): MonsterCache {
    return this.cache;
  }
}

/**
 * Cost asymmetry: ascent costs Bule, descent is free.
 *
 * Mirrors `Gnosis.PleromaticAsymmetryOfEffort`.
 */
export const CostLedger = {
  /**
   * Bule cost of ascending from `closureChain n` to `closureChain (n+1)`:
   * exactly `2 × closureChain n`. Grows as `2 × 10 × 3^n`.
   */
  ascentCost(closureValue: number): number {
    return 2 * closureValue;
  },

  /** Descent via fold is free. */
  descentCost(): 0 {
    return 0;
  },

  /** Cumulative ascent cost from grounding (10) to a target closure value. */
  cumulativeAscentCost(targetClosure: number): number {
    if (targetClosure <= PLEROMATIC_CLOSURE) return 0;
    return targetClosure - PLEROMATIC_CLOSURE;
  },
} as const;

/**
 * The grounding-triton as a structural object: 3 nodes, pairwise unit
 * equality, triple-point at (10, 1). Returns a snapshot witness.
 */
export function groundingTritonSnapshot(): {
  readonly nodes: readonly [GroundingNode, GroundingNode, GroundingNode];
  readonly tripleAgreement: boolean;
  readonly triplePoint: typeof TRIPLE_POINT;
} {
  const nodes = [
    GROUNDING_TRITON.math,
    GROUNDING_TRITON.physics,
    GROUNDING_TRITON.moonshine,
  ] as const;
  // Pairwise unit equality — mechanically rfl in Lean; runtime check here.
  const u0 = nodes[0].unit;
  const u1 = nodes[1].unit;
  const u2 = nodes[2].unit;
  const tripleAgreement = u0 === u1 && u1 === u2 && u0 === SHARED_UNIT;
  return { nodes, tripleAgreement, triplePoint: TRIPLE_POINT };
}
