// @a0n/.ignored_gnosis/distributed-inference-host/src/pleromatic-lensing.ts
/**
 * # Pleromatic Lensing Effect — Compression vs Lossless Retrieval
 *
 * Mirrors the Lean module `Gnosis.PleromaticLensingEffect`. The
 * Triton-stretch lens has *dual character*:
 *
 * * **Forward (descending) view, no index**: 9 Trihexenneon points
 *   collapse to 1 Hexon point (compression — many-to-one).
 * * **Backward (ascending) view, with copy-index**: each Trihexenneon
 *   position decomposes uniquely as `(hexonPosition, copyIndex)`,
 *   giving lossless retrieval. The copy-index is the residue mod 9.
 *
 * The cumulative Triton-stretch factor between Hexon (6) and
 * Trihexenneon (54) is `3 × 3 = 9` (two Triton walls). The lens
 * *folds* information rather than destroying it: position-only at the
 * higher level reappears as position-plus-residue at the lower level.
 *
 * ## Lean source-of-truth
 *
 * @see open-source/gnosis-math/Gnosis/PleromaticLensingEffect.lean
 * @see open-source/gnosis-math/Gnosis/PleromaticHorizonEffect.lean
 * @see open-source/gnosis-math/Gnosis/BraidedTower.lean
 */

/**
 * Cumulative two-stretch factor between Hexon and Trihexenneon
 * (`3 × 3 = 9`). Mirrors `two_triton_stretches_is_nine`.
 */
export const LENS_FACTOR_HEXON_TO_TRIHEXENNEON = 9 as const;

/**
 * Sanity check: Trihexenneon (54) = 9 × Hexon (6). Mirrors
 * `trihexenneon_is_nine_hexons`.
 */
export const TRIHEXENNEON_IS_NINE_HEXONS = 54 as const;

/**
 * The lossless decomposition of a Trihexenneon-position.
 *
 * `hexonPosition` is the higher-level (compressed) position;
 * `copyIndex ∈ {0,…,8}` is the residue that selects which of the 9
 * copies the original point lies in. Together they reconstruct the
 * original Trihexenneon-position perfectly.
 *
 * Mirrors the `(Nat × Nat)` pair returned by Lean `lensAddress`.
 */
export interface LensAddress {
  readonly hexonPosition: number;
  readonly copyIndex: number;
}

/**
 * Injectivity witness for the Triton-stretch: `3 * m === 3 * n`
 * implies `m === n`. Mirrors `triton_stretch_injective`. Returns the
 * truth-value of the implication for runtime assertion use.
 */
export function tritonStretchInjective(m: number, n: number): boolean {
  // If the antecedent fails, the implication is vacuously true.
  if (3 * m !== 3 * n) return true;
  return m === n;
}

/**
 * Division by 3 perfectly recovers the input from a Triton-stretch.
 * The *perfect-retrieval* witness for the upward lens. Mirrors
 * `triton_stretch_div_recovers`.
 */
export function tritonStretchDivRecovers(n: number): number {
  return Math.floor((3 * n) / 3);
}

/**
 * The compressing forward map: descend from Trihexenneon-position to
 * Hexon-position by dividing by 9. Many-to-one. Mirrors `lensDescent`.
 */
export function lensDescent(k: number): number {
  return Math.floor(k / LENS_FACTOR_HEXON_TO_TRIHEXENNEON);
}

/**
 * The addressed-descent map: returns both the Hexon-position and the
 * copy-index. The lossless reading of the lens. Mirrors `lensAddress`.
 */
export function lensAddress(k: number): LensAddress {
  return {
    hexonPosition: Math.floor(k / LENS_FACTOR_HEXON_TO_TRIHEXENNEON),
    copyIndex: k % LENS_FACTOR_HEXON_TO_TRIHEXENNEON,
  };
}

/**
 * Reconstruction map: rebuild the Trihexenneon-position from an
 * addressed descent. Mirrors `lensReconstruct`.
 */
export function lensReconstruct(addr: LensAddress): number {
  return LENS_FACTOR_HEXON_TO_TRIHEXENNEON * addr.hexonPosition + addr.copyIndex;
}

/**
 * **Lossless retrieval theorem** (one direction): decomposing then
 * reconstructing recovers the original position. Mirrors
 * `lens_address_reconstruct_roundtrip`. Returns the truth-value of
 * the equality for runtime assertion use.
 */
export function lensAddressReconstructRoundtrip(k: number): boolean {
  return lensReconstruct(lensAddress(k)) === k;
}

/**
 * The other direction: reconstructing then addressing recovers the
 * address, *provided* `r < 9`. Mirrors
 * `lens_reconstruct_address_roundtrip`. Returns the truth-value of
 * the equality for runtime assertion use.
 */
export function lensReconstructAddressRoundtrip(q: number, r: number): boolean {
  if (r >= LENS_FACTOR_HEXON_TO_TRIHEXENNEON) return false;
  const addr: LensAddress = { hexonPosition: q, copyIndex: r };
  const round = lensAddress(lensReconstruct(addr));
  return round.hexonPosition === q && round.copyIndex === r;
}

/**
 * Bule cost preservation under lensing: a unit step at Trihexenneon
 * resolution costs +1 reconstructed at the Hexon-addressed level too.
 * Always returns 1. Mirrors `lens_preserves_unit_cost`.
 */
export function lensPreservesUnitCost(k: number): number {
  return lensReconstruct(lensAddress(k + 1)) - lensReconstruct(lensAddress(k));
}

/**
 * **Pleromatic Lensing Effect master**: bundles the cumulative factor,
 * three concrete sample addresses (`18`, `27`, `53`), and roundtrip
 * witnesses. Mirrors `pleromatic_lensing_effect_master`.
 */
export const LensingMaster: {
  readonly factor: typeof LENS_FACTOR_HEXON_TO_TRIHEXENNEON;
  readonly trihexenneonIsNineHexons: typeof TRIHEXENNEON_IS_NINE_HEXONS;
  readonly sampleAddresses: {
    readonly eighteen: LensAddress;
    readonly twentyseven: LensAddress;
    readonly fiftythree: LensAddress;
  };
  readonly roundtripWitnesses: {
    readonly addressReconstructAtZero: boolean;
    readonly addressReconstructAtFiftythree: boolean;
    readonly reconstructAddressAtFiveEight: boolean;
  };
} = {
  factor: LENS_FACTOR_HEXON_TO_TRIHEXENNEON,
  trihexenneonIsNineHexons: TRIHEXENNEON_IS_NINE_HEXONS,
  sampleAddresses: {
    eighteen: lensAddress(18),
    twentyseven: lensAddress(27),
    fiftythree: lensAddress(53),
  },
  roundtripWitnesses: {
    addressReconstructAtZero: lensAddressReconstructRoundtrip(0),
    addressReconstructAtFiftythree: lensAddressReconstructRoundtrip(53),
    reconstructAddressAtFiveEight: lensReconstructAddressRoundtrip(5, 8),
  },
} as const;
