// @a0n/.ignored_gnosis/distributed-inference-host/src/pleromatic-bandwidth.ts
/**
 * # Pleromatic Bandwidth — Carrier-Bandwidth Asymmetry & Dual Phenomenology
 *
 * Mirrors two related Lean modules that both concern carrier bandwidth:
 *
 * 1. `Gnosis.PleromaticOneWayMirror` — the lensing asymmetry by carrier
 *    cardinality. The Triton's 3-position carrier cannot encode the
 *    Trihexenneon's 18-fold copy-index. The mirror is symbolic: lower
 *    observers lack the symbol-budget to register higher multiplicity,
 *    so structured higher signal arrives at the lower carrier as
 *    *steady collapse*, not noise.
 *
 * 2. `Gnosis.PleromaticEvolutionResolution` — the dual phenomenology of
 *    a +1 Bule step. At a given bandwidth, every step is either
 *    **evolution** (`k+1 ≥ bandwidth` — forces new symbols) or
 *    **resolution** (`k+1 < bandwidth` — refines an existing address).
 *    The reading is observer-relative; the unit cost (+1) is invariant
 *    across both readings and across all bandwidths.
 *
 * Both modules turn on the `carrierBandwidth` primitive, so they share
 * one runtime file. Constants for Triton (3), Hexon (6), and
 * Trihexenneon (54) appear in both.
 *
 * ## Lean source-of-truth
 *
 * @see open-source/gnosis-math/Gnosis/PleromaticOneWayMirror.lean
 * @see open-source/gnosis-math/Gnosis/PleromaticEvolutionResolution.lean
 */

/** Triton bandwidth: 3 distinguishable positions (`towerPhaseCount [3]`). */
export const BANDWIDTH_TRITON = 3 as const;

/** Hexon bandwidth: 6 distinguishable positions (`towerPhaseCount [3, 2]`). */
export const BANDWIDTH_HEXON = 6 as const;

/** Trihexenneon bandwidth: 54 positions (`towerPhaseCount [3, 2, 3, 3]`). */
export const BANDWIDTH_TRIHEXENNEON = 54 as const;

/** The compression factor from Trihexenneon → Triton (54 / 3 = 18). */
const TRITON_COMPRESSION_FACTOR = 18 as const;

/* -------------------------------------------------------------------------- */
/* Carrier bandwidth (PleromaticOneWayMirror)                                 */
/* -------------------------------------------------------------------------- */

/**
 * The carrier bandwidth of a tower level: the number of distinguishable
 * positions. For a level with phase count `n`, the bandwidth is `n`.
 *
 * Mirrors `Gnosis.PleromaticOneWayMirror.carrierBandwidth`.
 */
export function carrierBandwidth(level: number): number {
  return level;
}

/**
 * Each Triton-stretch *triples* the carrier bandwidth. Runtime witness
 * of `triton_stretch_triples_bandwidth`:
 * `carrierBandwidth (3 * n) = 3 * carrierBandwidth n`.
 */
export function tritonStretchTriplesBandwidth(n: number): boolean {
  return carrierBandwidth(3 * n) === 3 * carrierBandwidth(n);
}

/**
 * The bandwidth gap is strict: 54 > 3. Mirrors `bandwidth_gap_strict`.
 */
export function bandwidthGapStrict(): boolean {
  return BANDWIDTH_TRIHEXENNEON > BANDWIDTH_TRITON;
}

/**
 * Descent map from Trihexenneon-position to Triton-position: divide by
 * 18 (the cumulative compression factor `3 × 2 × 3 = 18`). The lower
 * observer's reading of a higher position.
 *
 * Mirrors `Gnosis.PleromaticOneWayMirror.lensDownToTriton`.
 */
export function lensDownToTriton(k: number): number {
  return Math.floor(k / TRITON_COMPRESSION_FACTOR);
}

/**
 * Eighteen consecutive Trihexenneon positions descend to one Triton
 * position. For `a, b ∈ [0, 18)`, `lensDownToTriton(a) === lensDownToTriton(b)`.
 *
 * Mirrors `lower_observer_cannot_distinguish_first_triton`.
 */
export function lowerObserverCannotDistinguishFirstTriton(
  a: number,
  b: number,
): boolean {
  if (a < 0 || a >= TRITON_COMPRESSION_FACTOR) return false;
  if (b < 0 || b >= TRITON_COMPRESSION_FACTOR) return false;
  return lensDownToTriton(a) === lensDownToTriton(b);
}

/**
 * The three 18-position fibers tiling the Trihexenneon: positions
 * `0..17 → 0`, `18..35 → 1`, `36..53 → 2`. Returns sample boundary
 * witnesses for each fiber.
 *
 * Mirrors `lens_triton_three_fibers`.
 */
export function lensTritonThreeFibers(): {
  fiber0: number[];
  fiber1: number[];
  fiber2: number[];
} {
  return {
    fiber0: [lensDownToTriton(0), lensDownToTriton(17)],
    fiber1: [lensDownToTriton(18), lensDownToTriton(35)],
    fiber2: [lensDownToTriton(36), lensDownToTriton(53)],
  };
}

/**
 * The Trihexenneon-internal copy-index ranges over 18 values; the
 * Triton's 3-position carrier cannot encode it. Mirrors
 * `copy_index_exceeds_triton_bandwidth`.
 */
export function copyIndexExceedsTritonBandwidth(): boolean {
  return TRITON_COMPRESSION_FACTOR > BANDWIDTH_TRITON;
}

/**
 * Eighteen distinct higher positions (`0..17`) all collapse to the same
 * Triton position (0). The lower observer sees them as a single event.
 *
 * Mirrors `eighteen_to_one_collision`.
 */
export function eighteenToOneCollision(): boolean {
  const target = lensDownToTriton(0);
  for (let i = 1; i < TRITON_COMPRESSION_FACTOR; i++) {
    if (lensDownToTriton(i) !== target) return false;
  }
  return target === 0;
}

/* -------------------------------------------------------------------------- */
/* Evolution-Resolution (PleromaticEvolutionResolution)                       */
/* -------------------------------------------------------------------------- */

/**
 * A +1 step from position `k` is an **evolution step** at the given
 * bandwidth iff `k + 1 ≥ bandwidth`. The new position is unrepresentable
 * at the current carrier; the step forces expansion.
 *
 * Mirrors `Gnosis.PleromaticEvolutionResolution.IsEvolutionStep`.
 */
export function isEvolutionStep(bandwidth: number, k: number): boolean {
  return k + 1 >= bandwidth;
}

/**
 * A +1 step from position `k` is a **resolution step** at the given
 * bandwidth iff `k + 1 < bandwidth`. The new position is already
 * representable; the step refines the existing address.
 *
 * Mirrors `Gnosis.PleromaticEvolutionResolution.IsResolutionStep`.
 */
export function isResolutionStep(bandwidth: number, k: number): boolean {
  return k + 1 < bandwidth;
}

/**
 * Every +1 step is *at least one* of evolution or resolution. The two
 * readings cover the step space. Mirrors `step_either_or`.
 */
export function stepEitherOr(bandwidth: number, k: number): boolean {
  return isEvolutionStep(bandwidth, k) || isResolutionStep(bandwidth, k);
}

/**
 * Every +1 step is *never both* evolution and resolution at a given
 * bandwidth. The two readings partition the step space.
 *
 * Mirrors `step_not_both`.
 */
export function stepNotBoth(bandwidth: number, k: number): boolean {
  return !(isEvolutionStep(bandwidth, k) && isResolutionStep(bandwidth, k));
}

/**
 * For any positive bandwidth, the position `bandwidth - 1` is the
 * evolution-trigger: the last representable position, after which the
 * +1 step exhausts the carrier.
 *
 * Mirrors `evolution_trigger_at_bandwidth_minus_one`.
 */
export function evolutionTriggerAtBandwidthMinusOne(bandwidth: number): boolean {
  if (bandwidth <= 0) return false;
  return isEvolutionStep(bandwidth, bandwidth - 1);
}

/**
 * The Bule cost of any +1 step is exactly 1, regardless of whether it
 * is read as evolution or resolution. The unit-cost heartbeat is
 * invariant across observer bandwidth.
 *
 * Mirrors `Gnosis.PleromaticEvolutionResolution.buleStepCost`.
 */
export function buleStepCost(_bandwidth: number, _k: number): 1 {
  return 1;
}

/**
 * The Bule cost is invariant under change of bandwidth: different
 * observers, same heartbeat. Mirrors
 * `bule_step_cost_invariant_under_bandwidth_change`.
 */
export function buleStepCostInvariantUnderBandwidthChange(
  bw1: number,
  bw2: number,
  k: number,
): boolean {
  return buleStepCost(bw1, k) === buleStepCost(bw2, k);
}

/* -------------------------------------------------------------------------- */
/* Master bundles                                                             */
/* -------------------------------------------------------------------------- */

/**
 * **One-Way Mirror master**: the asymmetric-by-cardinality bundle.
 * Strict bandwidth gap, 18-fold compression to Triton, and the
 * copy-index-exceeds-bandwidth flag.
 *
 * Mirrors `pleromatic_one_way_mirror_master`.
 */
export const OneWayMirrorMaster = {
  bandwidthGapStrict: bandwidthGapStrict(),
  compressionFactor: TRITON_COMPRESSION_FACTOR,
  copyIndexExceedsTritonBandwidth: copyIndexExceedsTritonBandwidth(),
  eighteenToOneCollision: eighteenToOneCollision(),
  trihexenneonIs18xTriton:
    BANDWIDTH_TRIHEXENNEON === TRITON_COMPRESSION_FACTOR * BANDWIDTH_TRITON,
} as const;

/**
 * **Evolution-vs-Resolution master**: every +1 step is either evolution
 * or resolution (exclusive and exhaustive); same step has different
 * readings at different bandwidths; cost is invariantly +1.
 *
 * Mirrors `pleromatic_evolution_resolution_master`. Sample readings:
 * step 2→3 is evolution at Triton (3) and resolution at Trihexenneon
 * (54); step 5→6 is evolution at Hexon (6) and resolution at
 * Trihexenneon (54).
 */
export const EvolutionResolutionMaster = {
  exclusivityAtTritonStep2: stepNotBoth(BANDWIDTH_TRITON, 2),
  exhaustivenessAtTritonStep2: stepEitherOr(BANDWIDTH_TRITON, 2),
  triton_step_two_three_isEvolution: isEvolutionStep(BANDWIDTH_TRITON, 2),
  trihexenneon_step_two_three_isResolution: isResolutionStep(
    BANDWIDTH_TRIHEXENNEON,
    2,
  ),
  hexon_step_five_six_isEvolution: isEvolutionStep(BANDWIDTH_HEXON, 5),
  trihexenneon_step_five_six_isResolution: isResolutionStep(
    BANDWIDTH_TRIHEXENNEON,
    5,
  ),
  trihexenneon_step_53_54_isEvolution: isEvolutionStep(
    BANDWIDTH_TRIHEXENNEON,
    53,
  ),
  costAlwaysOne: buleStepCost(BANDWIDTH_TRITON, 0) === 1,
  costInvariantUnderBandwidth: buleStepCostInvariantUnderBandwidthChange(
    BANDWIDTH_TRITON,
    BANDWIDTH_TRIHEXENNEON,
    7,
  ),
} as const;
