// @a0n/.ignored_gnosis/distributed-inference-host/src/provenance-transitivity.ts
/**
 * # Provenance Transitivity — from pairwise matching to same-provenance clusters
 *
 * Mirrors `Gnosis.Mesh.ProvenanceTransitivity` from `open-source/gnosis-math`,
 * the Rustic Church carrier that upgrades the *pairwise* same-device matching of
 * `Gnosis.Mesh.MicrophoneFingerprint` to a *transitive* same-provenance relation.
 * The physical picture:
 *
 *  - every recording carries a fingerprint (a per-band gain vector), with
 *    distance `fpDistance` summing the pointwise saturating `adist`;
 *  - if two recordings each match a common REFERENCE fingerprint within tolerance
 *    `tol`, they share provenance — and the metric triangle inequality forces
 *    them to be close to EACH OTHER (`fpDistance(x, y) <= 2*tol`);
 *  - the payoff: to certify a cluster of `n` members against a reference you
 *    check `n` reference distances, and the triangle inequality guarantees the
 *    `O(n^2)` pairwise bounds for free. Membership transits THROUGH the
 *    reference.
 *
 * The HARD NEGATIVE is that provenance does NOT transit across *distinct*
 * references: two items matched to references separated by more than `2*tol` are
 * not forced close to each other, so the cluster relation refuses to merge two
 * different provenance clusters.
 *
 * ## What is (and is not) carried
 *
 * This module is the runtime promotion check for a same-provenance observation.
 * It carries ONLY the finite contract the Lean file proves:
 *
 *  - the saturating per-band gain metric (`adist`);
 *  - the fingerprint distance as the summed pointwise `adist`, with a large
 *    sentinel on a length mismatch (`fpDistance` / `mismatchPenalty`), reused
 *    from the microphone-fingerprint carrier;
 *  - same-provenance admission against a reference within tolerance
 *    (`admitSameProvenance`), with the distinct-reference hard negative folded
 *    into the withheld outcome;
 *  - the transitivity bound (`clusterWithinTwoTol`): two members each within
 *    `tol` of a reference are within `2*tol` of each other — the runtime
 *    assertion that the Lean `provenance_transitive` theorem fires.
 *
 * Real fingerprint estimation, reference selection, clustering thresholds, and
 * population statistics stay runtime/empirical obligations OUTSIDE this
 * certificate — they are not theorems and are not asserted here.
 *
 * The promotion check folds the contract into the same epistemic-outcome
 * vocabulary the sibling mesh carriers already speak
 * (`claim_admitted` / `observation_only` / `claim_withheld`), so a same-provenance
 * claim cannot be displayed until it has cleared the Lean-side match gate.
 *
 * ## Lean source-of-truth
 *
 * @see open-source/gnosis-math/Gnosis/Mesh/ProvenanceTransitivity.lean
 * @see open-source/gnosis-math/Gnosis/Mesh/MicrophoneFingerprint.lean (pairwise base)
 */

/* ------------------------------------------------------------------ */
/*  Per-band distance and fingerprints (reused from MicrophoneFingerprint) */
/* ------------------------------------------------------------------ */

/**
 * A microphone fingerprint is a list of per-band gains over a fixed window.
 * Gains are non-negative integers (saturating `Nat` arithmetic), matching the
 * Lean `Fingerprint`.
 */
export type Fingerprint = readonly number[];

/** Absolute difference on `Nat` via saturating subtraction (`|a - b|`). The
 *  per-band gain residual metric, shared with the mesh corridor carriers.
 *  Mirrors `ProvenanceTransitivity.adist`. */
export function adist(a: number, b: number): number {
  return Math.max(0, a - b) + Math.max(0, b - a);
}

/**
 * Sentinel returned on a length mismatch: large enough that no sane tolerance
 * admits it, so differently-windowed fingerprints are treated as non-matching.
 * Mirrors `ProvenanceTransitivity.mismatchPenalty`.
 */
export const mismatchPenalty = 1000000;

/**
 * The fingerprint distance: the sum of pointwise `adist` between aligned bands.
 * On a length mismatch it returns `mismatchPenalty` rather than attempting a
 * partial alignment. Mirrors `ProvenanceTransitivity.fpDistance`.
 */
export function fpDistance(x: Fingerprint, y: Fingerprint): number {
  if (x.length !== y.length) {
    return mismatchPenalty;
  }
  let sum = 0;
  for (let i = 0; i < x.length; i++) {
    sum += adist(x[i], y[i]);
  }
  return sum;
}

/* ------------------------------------------------------------------ */
/*  Runtime promotion check (epistemic outcome)                       */
/* ------------------------------------------------------------------ */

/**
 * Epistemic outcome of a same-provenance observation. Identical vocabulary to
 * the sibling mesh carriers (`./microphone-fingerprint`,
 * `./electrical-network-frequency`), kept as a local literal union so this
 * certificate has no import coupling:
 *
 *  - `claim_admitted`    — the candidate fingerprint matches the reference within
 *                          tolerance, so the same-provenance claim may be
 *                          promoted (it joins the reference's cluster).
 *  - `observation_only`  — reserved for a degenerate self-reference check: the
 *                          candidate IS the reference (zero distance) so the
 *                          match is vacuous and reported, never promoted to a
 *                          cross-recording provenance claim.
 *  - `claim_withheld`    — the candidate is separated from the reference by more
 *                          than tolerance (a distinct reference / different
 *                          provenance), so there is no same-provenance claim.
 */
export type ProvenanceEpistemicOutcome =
  | 'claim_admitted'
  | 'observation_only'
  | 'claim_withheld';

/** The reason a same-provenance claim was held back from promotion. */
export type ProvenanceLimitationClass =
  | 'self_reference'
  | 'distinct_reference'
  | 'none';

export interface ProvenanceAdmission {
  readonly outcome: ProvenanceEpistemicOutcome;
  readonly limitation: ProvenanceLimitationClass;
  /** Whether the candidate cleared the distance tolerance against the reference. */
  readonly matched: boolean;
  /** The measured fingerprint distance to the reference (displayable separation). */
  readonly distance: number;
  readonly summary: string;
}

export interface ProvenanceAdmitOptions {
  /** Tolerance for the pointwise fingerprint distance match to the reference. */
  readonly tol: number;
}

/**
 * The promotion check Monster Mesh runs before any UI surface displays a
 * "this recording shares provenance with that reference" claim. It folds the
 * Lean contract into an epistemic outcome:
 *
 *  - the match gate decides whether the candidate shares the reference's
 *    provenance — a separation beyond tolerance is a distinct reference and is
 *    withheld (mirroring `distinct_reference_not_shared`);
 *  - a candidate matching within tolerance is admitted into the reference's
 *    cluster (mirroring `x_matches_ref` / `cluster_admitted`);
 *  - the degenerate self-reference case (candidate equal to the reference, zero
 *    distance) is a vacuous match: present but uncertifiable as a *cross*-
 *    recording provenance claim, so it lands as `observation_only`.
 *
 * @param x   the candidate recording's fingerprint (whose provenance is asserted)
 * @param ref the cluster reference fingerprint
 * @param opts the distance tolerance
 */
export function admitSameProvenance(
  x: Fingerprint,
  ref: Fingerprint,
  opts: ProvenanceAdmitOptions,
): ProvenanceAdmission {
  const distance = fpDistance(x, ref);
  const matched = distance <= opts.tol;

  // Self-reference gate: a candidate identical to the reference is a vacuous
  // match (distance 0). It is reported, never promoted to a cross-recording
  // provenance claim.
  if (distance === 0) {
    return {
      outcome: 'observation_only',
      limitation: 'self_reference',
      matched,
      distance,
      summary:
        'candidate is identical to the reference (distance 0) — vacuous self-match, not a cross-recording provenance claim',
    };
  }

  // Match gate: a candidate separated by more than tolerance belongs to a
  // distinct reference, so there is no same-provenance claim.
  if (!matched) {
    return {
      outcome: 'claim_withheld',
      limitation: 'distinct_reference',
      matched,
      distance,
      summary: `candidate separated from reference by ${distance} > tol ${opts.tol} — distinct provenance, not the same cluster`,
    };
  }

  return {
    outcome: 'claim_admitted',
    limitation: 'none',
    matched,
    distance,
    summary: `candidate matches the reference within tolerance (distance ${distance} <= ${opts.tol}) — same provenance, joins the cluster`,
  };
}

/* ------------------------------------------------------------------ */
/*  Transitivity bound helper                                         */
/* ------------------------------------------------------------------ */

/** The result of a transitivity-bound check between two cluster members. */
export interface TransitivityBound {
  /** Whether `x` matched the reference within `tol`. */
  readonly xMatchesRef: boolean;
  /** Whether `y` matched the reference within `tol`. */
  readonly yMatchesRef: boolean;
  /** Whether both members share the reference's provenance. */
  readonly bothMatch: boolean;
  /** The measured pairwise distance `fpDistance(x, y)`. */
  readonly pairDistance: number;
  /** The transitivity ceiling `2*tol`. */
  readonly bound: number;
  /** Whether the pairwise distance respects the `<= 2*tol` transitivity bound. */
  readonly withinBound: boolean;
}

/**
 * Demonstrates the transitivity bound the Lean `provenance_transitive` theorem
 * proves: if `x` and `y` each match a common reference within `tol`, then they
 * are within `2*tol` of each other. This is the metric triangle inequality at
 * work — the runtime mirror of the theorem firing.
 *
 * When both members share the reference (the precondition of the theorem) the
 * function ASSERTS the pairwise bound holds, throwing if a violation is ever
 * observed (which would mean the runtime metric disagrees with the proof). When
 * the precondition is not met (one member does not match the reference) it
 * reports the facts without asserting — the theorem says nothing there.
 *
 * @param x   first cluster member fingerprint
 * @param y   second cluster member fingerprint
 * @param ref the common reference fingerprint
 * @param tol the per-member tolerance to the reference
 */
export function clusterWithinTwoTol(
  x: Fingerprint,
  y: Fingerprint,
  ref: Fingerprint,
  tol: number,
): TransitivityBound {
  const xMatchesRef = fpDistance(x, ref) <= tol;
  const yMatchesRef = fpDistance(y, ref) <= tol;
  const bothMatch = xMatchesRef && yMatchesRef;
  const pairDistance = fpDistance(x, y);
  const bound = 2 * tol;
  const withinBound = pairDistance <= bound;

  // The transitivity theorem's guarantee: if both share the reference within
  // tol, the pairwise distance MUST respect 2*tol. Surface any violation loudly.
  if (bothMatch && !withinBound) {
    throw new Error(
      `provenance transitivity violated: x and y both match ref within ${tol} ` +
        `but fpDistance(x, y) = ${pairDistance} > 2*tol = ${bound}`,
    );
  }

  return { xMatchesRef, yMatchesRef, bothMatch, pairDistance, bound, withinBound };
}

/* ------------------------------------------------------------------ */
/*  Concrete fixtures (mirror the Lean fixtures)                      */
/* ------------------------------------------------------------------ */

/** The cluster reference fingerprint. Mirrors `ProvenanceTransitivity.ref`. */
export const ref: Fingerprint = [20, 20, 20];

/** Member `x`: `fpDistance(x, ref) = 1 + 1 + 0 = 2 <= tol(4)`. Matches the
 *  reference. Mirrors `ProvenanceTransitivity.x`. */
export const x: Fingerprint = [21, 19, 20];

/** Member `y`: `fpDistance(y, ref) = 1 + 1 + 1 = 3 <= tol(4)`. Matches the
 *  reference. By transitivity `fpDistance(x, y) = 2 + 2 + 1 = 5 <= 2*tol = 8`.
 *  Mirrors `ProvenanceTransitivity.y`. */
export const y: Fingerprint = [19, 21, 21];

/** A far item `z`: `fpDistance(z, ref) = 20 + 20 + 20 = 60`, well beyond
 *  tolerance 4. Does NOT match the reference. Mirrors `ProvenanceTransitivity.z`. */
export const z: Fingerprint = [40, 40, 40];

/** A DISTINCT reference `ref2`, far from `ref`: `fpDistance(ref2, ref) = 60`,
 *  well beyond `2*tol = 8`. Mirrors `ProvenanceTransitivity.ref2`. */
export const ref2: Fingerprint = [40, 40, 40];

/** Member `w` of the OTHER cluster: `fpDistance(w, ref2) = 1 + 1 + 0 = 2 <=
 *  tol(4)`. It matches `ref2`, NOT `ref`. Mirrors `ProvenanceTransitivity.w`. */
export const w: Fingerprint = [41, 39, 40];

/**
 * Provenance Transitivity master witness. Bundles the boundary facts the Lean
 * theorem `provenance_transitivity_master` proves, each recomputed here from
 * first principles (never imported from Lean IR).
 *
 * Mirrors `ProvenanceTransitivity.provenance_transitivity_master`.
 */
export const ProvenanceTransitivityMaster = {
  /** `x` matches the reference within tolerance 4 (distance 2). */
  xMatchesRef:
    admitSameProvenance(x, ref, { tol: 4 }).outcome === 'claim_admitted',

  /** `y` matches the reference within tolerance 4 (distance 3). */
  yMatchesRef:
    admitSameProvenance(y, ref, { tol: 4 }).outcome === 'claim_admitted',

  /** The far item `z` does NOT match the reference within tolerance 4 (distance 60). */
  zNotMatchRef:
    admitSameProvenance(z, ref, { tol: 4 }).outcome === 'claim_withheld',

  /** The transitivity bound fires: `fpDistance(x, y) = 5 <= 2*tol = 8`. */
  transitivityBoundHolds: clusterWithinTwoTol(x, y, ref, 4).withinBound,

  /** And the precondition genuinely holds (both members share the reference). */
  clusterBothMatch: clusterWithinTwoTol(x, y, ref, 4).bothMatch,

  /** The two references really are separated by more than `2*tol` (distance 60 > 8). */
  referencesWellSeparated: fpDistance(ref2, ref) > 2 * 4,

  /** Hard negative: `x` (provenance `ref`) and `w` (provenance `ref2`) are NOT
   *  within `2*tol` of each other (`fpDistance(x, w) = 60 > 8`). Provenance does
   *  not transit across distinct references. */
  distinctReferenceNotShared: !(fpDistance(x, w) <= 2 * 4),
} as const;
