// @a0n/.ignored_gnosis/distributed-inference-host/src/qwen-tokenizer.ts
import { closeSync, openSync, readFileSync, readSync } from 'node:fs';

const GGUF_MAGIC = 0x46554747;

const GGUF_TYPES = {
  UINT8: 0,
  INT8: 1,
  UINT16: 2,
  INT16: 3,
  UINT32: 4,
  INT32: 5,
  FLOAT32: 6,
  BOOL: 7,
  STRING: 8,
  ARRAY: 9,
  UINT64: 10,
  INT64: 11,
  FLOAT64: 12,
} as const;

const QWEN_PRETOKENIZER_PATTERN =
  /'s|'t|'re|'ve|'m|'ll|'d|[^\r\n\p{L}\p{N}]?\p{L}+|\p{N}| ?[^\s\p{L}\p{N}]+[\r\n]*|\s*[\r\n]+|\s+(?!\S)|\s+/giu;

export interface QwenTokenizerMetadata {
  tokens: string[];
  merges: string[];
  bosTokenId?: number;
  eosTokenId?: number;
  paddingTokenId?: number;
}

type HuggingFaceTokenizerJson = {
  model?: {
    type?: string;
    vocab?: Record<string, number>;
    merges?: string[] | Array<[string, string]>;
  };
  added_tokens?: Array<{
    id?: number;
    content?: string;
    special?: boolean;
  }>;
  bos_token?: string | { id?: number; content?: string };
  eos_token?: string | { id?: number; content?: string };
  pad_token?: string | { id?: number; content?: string };
};

type ByteMaps = {
  byteEncoder: string[];
  byteDecoder: Map<string, number>;
};

const TEXT_ENCODER = new TextEncoder();
const TEXT_DECODER = new TextDecoder('utf-8', { fatal: false });

function buildByteMaps(): ByteMaps {
  const bytes: number[] = [];
  for (let i = '!'.charCodeAt(0); i <= '~'.charCodeAt(0); i++) bytes.push(i);
  for (let i = 0xa1; i <= 0xac; i++) bytes.push(i);
  for (let i = 0xae; i <= 0xff; i++) bytes.push(i);

  const chars = [...bytes];
  let next = 0;
  for (let byte = 0; byte < 256; byte++) {
    if (bytes.includes(byte)) continue;
    bytes.push(byte);
    chars.push(256 + next);
    next += 1;
  }

  const byteEncoder = Array<string>(256);
  const byteDecoder = new Map<string, number>();
  for (let i = 0; i < bytes.length; i++) {
    const char = String.fromCharCode(chars[i]!);
    byteEncoder[bytes[i]!] = char;
    byteDecoder.set(char, bytes[i]!);
  }

  return { byteEncoder, byteDecoder };
}

const BYTE_MAPS = buildByteMaps();

class GGUFMetadataReader {
  private fd: number;
  private offset = 0;
  private scratch = Buffer.alloc(1024);

  constructor(filePath: string) {
    this.fd = openSync(filePath, 'r');
  }

  close(): void {
    closeSync(this.fd);
  }

  private read(size: number): Buffer {
    if (size > this.scratch.length) this.scratch = Buffer.alloc(size);
    const bytesRead = readSync(this.fd, this.scratch, 0, size, this.offset);
    if (bytesRead !== size) {
      throw new Error(`unexpected EOF while reading GGUF metadata`);
    }
    this.offset += size;
    return this.scratch.subarray(0, size);
  }

  private skip(size: number): void {
    this.offset += size;
  }

  private readUint32(): number {
    return this.read(4).readUInt32LE(0);
  }

  private readInt32(): number {
    return this.read(4).readInt32LE(0);
  }

  private readUint64(): number {
    return Number(this.read(8).readBigUInt64LE(0));
  }

  private readString(): string {
    const length = this.readUint64();
    const buffer = Buffer.alloc(length);
    const bytesRead = readSync(this.fd, buffer, 0, length, this.offset);
    if (bytesRead !== length) {
      throw new Error(`unexpected EOF while reading GGUF string`);
    }
    this.offset += length;
    return buffer.toString('utf8');
  }

  private skipString(): void {
    const length = this.readUint64();
    this.skip(length);
  }

  private skipValue(type: number): void {
    switch (type) {
      case GGUF_TYPES.UINT8:
      case GGUF_TYPES.INT8:
      case GGUF_TYPES.BOOL:
        this.skip(1);
        return;
      case GGUF_TYPES.UINT16:
      case GGUF_TYPES.INT16:
        this.skip(2);
        return;
      case GGUF_TYPES.UINT32:
      case GGUF_TYPES.INT32:
      case GGUF_TYPES.FLOAT32:
        this.skip(4);
        return;
      case GGUF_TYPES.UINT64:
      case GGUF_TYPES.INT64:
      case GGUF_TYPES.FLOAT64:
        this.skip(8);
        return;
      case GGUF_TYPES.STRING:
        this.skipString();
        return;
      case GGUF_TYPES.ARRAY: {
        const arrayType = this.readUint32();
        const arrayLength = this.readUint64();
        for (let i = 0; i < arrayLength; i++) {
          this.skipValue(arrayType);
        }
        return;
      }
      default:
        throw new Error(`unsupported GGUF metadata type: ${type}`);
    }
  }

  private readStringArray(): string[] {
    const arrayType = this.readUint32();
    const arrayLength = this.readUint64();
    if (arrayType !== GGUF_TYPES.STRING) {
      throw new Error(`expected GGUF string array, got type ${arrayType}`);
    }

    const values: string[] = [];
    for (let i = 0; i < arrayLength; i++) {
      values.push(this.readString());
    }
    return values;
  }

  readTokenizerMetadata(): QwenTokenizerMetadata {
    const magic = this.readUint32();
    if (magic !== GGUF_MAGIC) {
      throw new Error(`invalid GGUF magic: ${magic}`);
    }

    const version = this.readUint32();
    if (version < 2) {
      throw new Error(`unsupported GGUF version: ${version}`);
    }

    this.readUint64();
    const metadataCount = this.readUint64();
    const metadata: QwenTokenizerMetadata = {
      tokens: [],
      merges: [],
    };

    for (let i = 0; i < metadataCount; i++) {
      const key = this.readString();
      const valueType = this.readUint32();

      if (key === 'tokenizer.ggml.tokens') {
        metadata.tokens = this.readStringArray();
      } else if (key === 'tokenizer.ggml.merges') {
        metadata.merges = this.readStringArray();
      } else if (key === 'tokenizer.ggml.bos_token_id') {
        metadata.bosTokenId = this.readUint32();
      } else if (key === 'tokenizer.ggml.eos_token_id') {
        metadata.eosTokenId = this.readUint32();
      } else if (key === 'tokenizer.ggml.padding_token_id') {
        metadata.paddingTokenId = this.readUint32();
      } else if (valueType === GGUF_TYPES.INT32 && key.endsWith('_token_id')) {
        this.readInt32();
      } else if (valueType === GGUF_TYPES.UINT32 && key.endsWith('_token_id')) {
        this.readUint32();
      } else {
        this.skipValue(valueType);
      }
    }

    if (metadata.tokens.length === 0) {
      throw new Error('GGUF metadata does not contain tokenizer.ggml.tokens');
    }
    if (metadata.merges.length === 0) {
      throw new Error('GGUF metadata does not contain tokenizer.ggml.merges');
    }

    return metadata;
  }
}

function pairKey(left: string, right: string): string {
  return `${left}\u0000${right}`;
}

function byteEncode(text: string): string {
  let encoded = '';
  for (const byte of TEXT_ENCODER.encode(text)) {
    encoded += BYTE_MAPS.byteEncoder[byte]!;
  }
  return encoded;
}

function byteDecode(encoded: string): string {
  const bytes: number[] = [];
  for (const char of Array.from(encoded)) {
    const byte = BYTE_MAPS.byteDecoder.get(char);
    if (byte !== undefined) {
      bytes.push(byte);
      continue;
    }

    for (const fallbackByte of TEXT_ENCODER.encode(char)) {
      bytes.push(fallbackByte);
    }
  }

  return TEXT_DECODER.decode(new Uint8Array(bytes));
}

export class QwenTokenizer {
  private readonly tokenToId: Map<string, number>;
  private readonly idToToken: string[];
  private readonly mergeRanks: Map<string, number>;
  private readonly specialTokens: Array<{ token: string; id: number }>;
  readonly stopTokenIds: number[];

  constructor(metadata: QwenTokenizerMetadata) {
    this.idToToken = metadata.tokens;
    this.tokenToId = new Map(
      metadata.tokens.map((token, id) => [token, id] as const)
    );
    this.mergeRanks = new Map();
    for (let rank = 0; rank < metadata.merges.length; rank++) {
      const merge = metadata.merges[rank]!;
      const split = merge.indexOf(' ');
      if (split < 0) continue;
      this.mergeRanks.set(
        pairKey(merge.slice(0, split), merge.slice(split + 1)),
        rank
      );
    }

    this.specialTokens = metadata.tokens
      .map((token, id) => ({ token, id }))
      .filter(({ token }) =>
        token.startsWith('<|') ||
        token === '<tool_call>' ||
        token === '</tool_call>' ||
        token.startsWith('[PAD') ||
        token.startsWith('</')
      )
      .sort((a, b) => b.token.length - a.token.length);

    this.stopTokenIds = [
      metadata.eosTokenId,
      metadata.paddingTokenId,
      metadata.bosTokenId,
    ].filter((id): id is number => typeof id === 'number');
  }

  encode(text: string): number[] {
    const tokenIds: number[] = [];
    let cursor = 0;

    while (cursor < text.length) {
      const matchedSpecial = this.specialTokens.find(({ token }) =>
        text.startsWith(token, cursor)
      );
      if (matchedSpecial) {
        tokenIds.push(matchedSpecial.id);
        cursor += matchedSpecial.token.length;
        continue;
      }

      let nextSpecialIndex = text.length;
      for (const { token } of this.specialTokens) {
        const index = text.indexOf(token, cursor);
        if (index >= 0 && index < nextSpecialIndex) {
          nextSpecialIndex = index;
        }
      }

      tokenIds.push(...this.encodeOrdinary(text.slice(cursor, nextSpecialIndex)));
      cursor = nextSpecialIndex;
    }

    return tokenIds;
  }

  decode(tokenIds: readonly number[]): string {
    let encoded = '';
    for (const tokenId of tokenIds) {
      const token = this.idToToken[tokenId];
      if (!token || this.isSpecialToken(token)) continue;
      encoded += token;
    }

    return byteDecode(encoded);
  }

  private encodeOrdinary(text: string): number[] {
    const tokenIds: number[] = [];
    for (const match of text.matchAll(QWEN_PRETOKENIZER_PATTERN)) {
      const piece = byteEncode(match[0]);
      const directId = this.tokenToId.get(piece);
      if (directId !== undefined) {
        tokenIds.push(directId);
        continue;
      }

      for (const token of this.bpe(piece)) {
        const tokenId = this.tokenToId.get(token);
        if (tokenId === undefined) {
          throw new Error(`Qwen tokenizer missing BPE token: ${token}`);
        }
        tokenIds.push(tokenId);
      }
    }
    return tokenIds;
  }

  private bpe(piece: string): string[] {
    let parts = Array.from(piece);
    if (parts.length <= 1) return parts;

    while (parts.length > 1) {
      let bestRank = Number.POSITIVE_INFINITY;
      let bestLeft = '';
      let bestRight = '';

      for (let i = 0; i < parts.length - 1; i++) {
        const rank = this.mergeRanks.get(pairKey(parts[i]!, parts[i + 1]!));
        if (rank !== undefined && rank < bestRank) {
          bestRank = rank;
          bestLeft = parts[i]!;
          bestRight = parts[i + 1]!;
        }
      }

      if (!Number.isFinite(bestRank)) break;

      const merged: string[] = [];
      for (let i = 0; i < parts.length; i++) {
        if (i < parts.length - 1 && parts[i] === bestLeft && parts[i + 1] === bestRight) {
          merged.push(bestLeft + bestRight);
          i += 1;
        } else {
          merged.push(parts[i]!);
        }
      }
      parts = merged;
    }

    return parts;
  }

  private isSpecialToken(token: string): boolean {
    return (
      token.startsWith('<|') ||
      token === '<tool_call>' ||
      token === '</tool_call>' ||
      token.startsWith('[PAD')
    );
  }
}

/**
 * Loads the Qwen Tokenizer From Gguf.
 */
export function loadQwenTokenizerFromGguf(ggufPath: string): QwenTokenizer {
  const reader = new GGUFMetadataReader(ggufPath);
  try {
    return new QwenTokenizer(reader.readTokenizerMetadata());
  } finally {
    reader.close();
  }
}

function normalizeMergeEntry(entry: string | [string, string]): string {
  return Array.isArray(entry) ? `${entry[0]} ${entry[1]}` : entry;
}

function tokenIdFromJsonToken(
  token: string | { id?: number; content?: string } | undefined,
  tokenToId: Map<string, number>,
): number | undefined {
  if (token == null) return undefined;
  if (typeof token === 'object' && token.id !== undefined) return token.id;
  const content = typeof token === 'string' ? token : token.content;
  return content === undefined ? undefined : tokenToId.get(content);
}

/**
 * Loads the Qwen Tokenizer Metadata From Json.
 */
export function loadQwenTokenizerMetadataFromJson(jsonPath: string): QwenTokenizerMetadata {
  const parsed = JSON.parse(readFileSync(jsonPath, 'utf8')) as HuggingFaceTokenizerJson;
  const vocab = parsed.model?.vocab;
  if (!vocab || typeof vocab !== 'object') {
    throw new Error(`Qwen tokenizer JSON is missing model.vocab`);
  }

  const tokenToId = new Map<string, number>();
  let maxTokenId = -1;
  for (const [token, id] of Object.entries(vocab)) {
    if (!Number.isInteger(id) || id < 0) {
      throw new Error(`Qwen tokenizer JSON has invalid token id for ${JSON.stringify(token)}`);
    }
    tokenToId.set(token, id);
    maxTokenId = Math.max(maxTokenId, id);
  }

  for (const addedToken of parsed.added_tokens ?? []) {
    if (
      addedToken.content === undefined ||
      addedToken.id === undefined ||
      !Number.isInteger(addedToken.id) ||
      addedToken.id < 0
    ) {
      continue;
    }
    tokenToId.set(addedToken.content, addedToken.id);
    maxTokenId = Math.max(maxTokenId, addedToken.id);
  }

  const tokens = Array<string>(maxTokenId + 1);
  for (const [token, id] of tokenToId) {
    tokens[id] = token;
  }
  for (let id = 0; id < tokens.length; id++) {
    if (tokens[id] === undefined) {
      throw new Error(`Qwen tokenizer JSON has no token for id ${id}`);
    }
  }

  const merges = (parsed.model?.merges ?? []).map(normalizeMergeEntry);
  const eosTokenId =
    tokenIdFromJsonToken(parsed.eos_token, tokenToId) ??
    tokenToId.get('<|im_end|>') ??
    tokenToId.get('<|endoftext|>');

  const metadata: QwenTokenizerMetadata = {
    tokens,
    merges,
  };
  const bosTokenId = tokenIdFromJsonToken(parsed.bos_token, tokenToId);
  if (bosTokenId !== undefined) metadata.bosTokenId = bosTokenId;
  if (eosTokenId !== undefined) metadata.eosTokenId = eosTokenId;
  const paddingTokenId =
    tokenIdFromJsonToken(parsed.pad_token, tokenToId) ?? tokenToId.get('<|endoftext|>');
  if (paddingTokenId !== undefined) metadata.paddingTokenId = paddingTokenId;
  return metadata;
}

/**
 * Loads the Qwen Tokenizer From Json.
 */
export function loadQwenTokenizerFromJson(jsonPath: string): QwenTokenizer {
  return new QwenTokenizer(loadQwenTokenizerMetadataFromJson(jsonPath));
}
