// @a0n/.ignored_gnosis/distributed-inference-host/src/semantic-sieve-cache.ts
//! Sovereign-sieve SEMANTIC cache lane for the transformer serving host.
//!
//! This is the host (Cloud Run / Node) analog of `apps/ssm-fleet/src/semantic-cache.ts`
//! (the edge, R2-backed, worker-isolate reference). It wires the model-agnostic,
//! Lean-proven primitive `open-source/bitwise/sovereign-sieve-key.ts` into the
//! transformer mesh (mistral-7b, qwen, llama fat-station knots) so the fleet can
//! serve a PARAPHRASE of a prior query with zero decode.
//!
//! WHERE THIS SITS: strictly ABOVE the existing exact-match global cache
//! (`skymesh-cache.ts` — fp48 over the prompt TOKENS, teleports another node's
//! answer only on a byte-identical prompt). The sieve catches the SEMANTIC
//! near-duplicate the token hash misses: two prompts whose final pre-lm_head
//! hidden states are cosine-close (τ ≥ 0.72). It composes, it does not conflict —
//! the exact cache is checked first (cheapest), the sieve second, decode last.
//!
//! THE KEY: the prompt's final pre-lm_head hidden = the transformer's post-all-layers
//! tail residual of the last prompt token (shape [hiddenDim]). THE FOLD (landed):
//! `Pipeline.generateWithPromptHidden` issues ONE native `/generate` with
//! `return_hidden: true` (fat-station.rs `prompt_hidden`, commit 02618367) so a single
//! forward serves BOTH this key AND the first-K decode tokens for the agreement gate —
//! no separate host-orchestrated prefill. `Pipeline.computePromptHidden` (the old
//! separate-prefill accessor) stays as the multi-station / host-orchestrated fallback.
//!
//! SAFETY: a bucket collision that fails cosine-verify is a MISS, never a wrong
//! answer (miss-not-lie). On top of that, the CALLER applies a DECODE-AGREEMENT
//! gate (the first K decoded tokens must equal the cached answer's first K) before
//! serving — the entity-safety proven essential on the edge fleet. This module
//! owns the sieve + persistence; the decode-agreement gate lives in the serve lane.
//!
//! STORAGE: local disk (per-model sieve JSON blob), mirroring the disk-backed
//! amplituhedron prompt_cache the host already uses (NOT R2 — the host is a Node
//! process, not a Worker isolate with an R2 binding). Cross-instance sharing on
//! Cloud Run is a documented follow-up (an edgework blob endpoint, the way
//! skymesh-cache.ts reaches edgework over HTTP). Writes are COALESCED so default-on
//! learning never write-amplifies.

import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { SovereignSieve, type SieveBlob } from '../../../bitwise/sovereign-sieve-key';

export interface CachedAnswer {
  text: string;
  tokens: number[];
}

const MAX_SEEDS = 2000; // bound the blob; evict oldest beyond this (LRU-ish by insertion)
const FLUSH_EVERY_SEEDS = 8; // coalesce up to N misses into ONE disk write (learning stays cheap)
const FLUSH_EVERY_MS = 20_000; // ...or flush at most every 20s, whichever comes first
const DEFAULT_TAU = 0.72; // the measured 100%-hit / 0%-miss operating point (rwkv7-0.4b activations)

// ── Config (default-OFF; arm explicitly, mirroring the guarded roll-out contract) ──

/** OFF unless MOONSHINE_SOVEREIGN_SIEVE is '1'/'true'/'on'. Starts guarded (the edge
 *  fleet proved default-on-ABLE, but this transformer wiring ships behind the flag
 *  until it is deployed + measured on the GPU mesh). Kill switch = unset/0. */
export function sovereignSieveEnabled(): boolean {
  const raw = process.env['MOONSHINE_SOVEREIGN_SIEVE'];
  if (raw === undefined) return false;
  const s = raw.trim().toLowerCase();
  return s === '1' || s === 'true' || s === 'on';
}

/** Decode-agreement K: the number of leading decoded tokens that must match the
 *  cached answer before we serve it. Higher = safer + more confirm-decode cost. */
export function sieveAgreementK(): number {
  const raw = Number(process.env['MOONSHINE_SOVEREIGN_SIEVE_AGREEMENT_K'] ?? 8);
  return Number.isInteger(raw) && raw > 0 ? raw : 8;
}

function sieveTau(): number {
  const raw = Number(process.env['MOONSHINE_SOVEREIGN_SIEVE_TAU'] ?? DEFAULT_TAU);
  return Number.isFinite(raw) && raw > 0 && raw <= 1 ? raw : DEFAULT_TAU;
}

/** Per-model sieve file. The hidden states are WEIGHT-SPECIFIC (a residual under one
 *  model must never key another), so the sieve is model-scoped — never model-agnostic
 *  like the exact-match global cache. */
function sieveDir(): string {
  return process.env['MOONSHINE_SOVEREIGN_SIEVE_DIR'] || join(process.cwd(), '.sovereign-sieve');
}
function blobPath(model: string): string {
  const safe = (model || 'unknown').replace(/[^a-zA-Z0-9._-]/g, '_');
  return join(sieveDir(), `${safe}.sieve.json`);
}

// ── Per-process (per-instance) state: rehydrate the sieve once, like the resident engine ──

const sieveByModel = new Map<string, SovereignSieve<CachedAnswer>>();
const pendingByModel = new Map<string, number>();
const lastFlushByModel = new Map<string, number>();

function loadSieve(model: string): SovereignSieve<CachedAnswer> {
  const existing = sieveByModel.get(model);
  if (existing) return existing;
  let sieve: SovereignSieve<CachedAnswer>;
  try {
    const path = blobPath(model);
    if (existsSync(path)) {
      const blob = JSON.parse(readFileSync(path, 'utf8')) as SieveBlob<CachedAnswer>;
      sieve = SovereignSieve.fromBlob<CachedAnswer>(blob);
    } else {
      sieve = new SovereignSieve<CachedAnswer>({ tau: sieveTau() });
    }
  } catch {
    sieve = new SovereignSieve<CachedAnswer>({ tau: sieveTau() });
  }
  sieveByModel.set(model, sieve);
  return sieve;
}

/** Verified semantic lookup on the prompt's final hidden. Returns the cached
 *  completion on a hit (candidate for the decode-agreement gate), else null. */
export function semanticLookup(
  model: string,
  hidden: Float32Array,
): { answer: CachedAnswer; cosine: number; candidates: number } | null {
  const sieve = loadSieve(model);
  if (sieve.size === 0) return null;
  return sieve.lookup(hidden);
}

/** Learn a freshly-computed completion. Added to the in-memory sieve IMMEDIATELY
 *  (serves the next request in THIS instance); the disk flush is COALESCED — every N
 *  seeds or T seconds, not per-miss — so default-on learning never write-amplifies.
 *  Seeds since the last flush are lost if the instance dies first; they are cheaply
 *  re-learned. Best-effort: never throws (the serve path must not fail on a cache write). */
export function seedCompletion(model: string, hidden: Float32Array, answer: CachedAnswer): void {
  try {
    const sieve = loadSieve(model);
    sieve.seed(hidden, answer);
    const pending = (pendingByModel.get(model) ?? 0) + 1;
    pendingByModel.set(model, pending);
    const now = Date.now();
    const lastFlush = lastFlushByModel.get(model) ?? 0;
    // First seed (lastFlush=0) persists immediately (now ≫ 20s); later bursts within the
    // window coalesce, so an idle cache still persists while a burst is one write.
    if (pending >= FLUSH_EVERY_SEEDS || now - lastFlush >= FLUSH_EVERY_MS) {
      pendingByModel.set(model, 0);
      lastFlushByModel.set(model, now);
      const blob = sieve.toBlob();
      if (blob.seeds.length > MAX_SEEDS) blob.seeds = blob.seeds.slice(blob.seeds.length - MAX_SEEDS);
      const path = blobPath(model);
      const dir = dirname(path);
      if (!existsSync(dir)) mkdirSync(dir, { recursive: true });
      writeFileSync(path, JSON.stringify(blob));
    }
  } catch {
    /* best-effort: a cache write must never break serving */
  }
}
