// @a0n/.ignored_gnosis/distributed-inference-host/src/sigma-attractor.ts
// sigma-attractor.ts
// Principled sigma selection for image deblur via the phi-attractor.
//
// Mapped onto FORK/RACE/FOLD:
//   FORK  — each phi-grid candidate's wienerDeblur + measureSpectrumBands
//           runs as a separate task (`forkRaceFold` from ./frf). On a single-
//           thread Node host async tasks don't parallelise CPU work, but the
//           shape matches the canonical pattern and the same structure
//           transparently parallelises behind worker_threads when wired up.
//   RACE  — identity (no filtering). All Wiener forks contribute to the fold.
//   FOLD  — the Fibonacci sliding window is INHERENTLY SEQUENTIAL (each
//           candidate's deficit ratio depends on the previous candidate's
//           energy), so the fold walks the per-candidate energies in input
//           order, updates the attractor state, and selects the argmin
//           |Binet residual|. A sqrt(5)-normalised Newton step against the
//           lower neighbour produces refinedSigma.
//
// Mirrors the deficit-ratio mechanic of interfere() at
// open-source/aether/src/glossolalia-moa.ts:160.

import {
  PHI,
  binetResidual,
  goldenSpacedGrid,
  createGoldenAttractorState,
  updateGoldenAttractor,
  phiAttractorNewtonStep,
  type GoldenAttractorState,
} from './attractor';
import { gaussianPSF } from './image-psf';
import { wienerDeblur } from './image-deblur';
import { measureSpectrumBands } from './image-spectrum-classify';
import { forkRaceFold, type ForkTask } from './frf';

export interface SelectSigmaArgs {
  plane: Float32Array;       // power-of-2-sized input (mirror-padded by caller)
  w: number;                 // power of 2
  h: number;                 // power of 2
  psfSize?: number;          // default 11; positive odd integer
  sigmaMin?: number;         // default 0.5
  candidateCount?: number;   // default 7 (covers sigma in [0.5, ~9])
  noiseFloor?: number;       // forwarded to wienerDeblur, default 1e-3
}

export interface SigmaCandidateReport {
  sigma: number;
  highBandEnergy: number;
  deficitRatio: number;            // 1.0 for cycle <= 1 (no ratio yet)
  binetResidual: number;           // (deficitRatio - PHI) / SQRT_5
  excludedFromAttractor: boolean;  // true for first sample (cycle <= 1)
}

export interface SelectSigmaResult {
  selectedSigma: number;                   // best grid candidate by min |residual|
  refinedSigma: number;                    // Newton step from best + neighbor; clamped
  candidates: ReadonlyArray<SigmaCandidateReport>;
  bestCandidateIndex: number;              // index into candidates[]
}

function isPositivePow2(n: number): boolean {
  return Number.isInteger(n) && n > 0 && (n & (n - 1)) === 0;
}

/**
 * Select a principled blur sigma by sweeping a phi-spaced grid of candidate
 * sigmas, deblurring with each (FORK), and folding the per-candidate
 * recovered high-band energies through the Fibonacci sliding-window
 * attractor (FOLD). The deficit ratio (last/prev) approaches phi at the
 * recovery curve's inflection point; the candidate minimising the absolute
 * Binet residual is chosen, then refined via a sqrt(5)-normalised Newton
 * step against its lower neighbour. Synchronous public surface — internally
 * uses `forkRaceFold` for the structural parallelism shape.
 */
export async function selectSigmaByGoldenAttractorAsync(
  args: SelectSigmaArgs,
): Promise<SelectSigmaResult> {
  const {
    plane,
    w,
    h,
    psfSize = 11,
    sigmaMin = 0.5,
    candidateCount = 7,
    noiseFloor = 1e-3,
  } = args;

  // 1. Validation.
  if (!isPositivePow2(w)) {
    throw new RangeError(`selectSigmaByGoldenAttractor: w must be a positive power of 2, got ${w}`);
  }
  if (!isPositivePow2(h)) {
    throw new RangeError(`selectSigmaByGoldenAttractor: h must be a positive power of 2, got ${h}`);
  }
  if (plane.length !== w * h) {
    throw new RangeError(
      `selectSigmaByGoldenAttractor: plane.length=${plane.length} != w*h=${w * h}`,
    );
  }
  if (!Number.isInteger(psfSize) || psfSize <= 0 || (psfSize & 1) === 0) {
    throw new RangeError(
      `selectSigmaByGoldenAttractor: psfSize must be a positive odd integer, got ${psfSize}`,
    );
  }
  if (!Number.isFinite(sigmaMin) || sigmaMin <= 0) {
    throw new RangeError(`selectSigmaByGoldenAttractor: sigmaMin must be > 0, got ${sigmaMin}`);
  }
  if (!Number.isInteger(candidateCount) || candidateCount < 3) {
    throw new RangeError(
      `selectSigmaByGoldenAttractor: candidateCount must be an integer >= 3, got ${candidateCount}`,
    );
  }

  // 2. Phi-spaced sigma grid.
  const grid = goldenSpacedGrid(sigmaMin, candidateCount);

  // 3. FORK: per-candidate Wiener + spectral bands. forkRaceFold preserves
  //    fork-input order (NOT completion order) so the Fibonacci fold below
  //    sees energies in sigma-ascending order — critical for the deficit-
  //    ratio recursion.
  type CandidateEnergy = { sigma: number; highBandEnergy: number };
  const forks: ForkTask<CandidateEnergy>[] = grid.map((sigma) => async () => {
    const psf = gaussianPSF(sigma, psfSize);
    const recovered = wienerDeblur({
      plane, w, h, psf, psfW: psfSize, psfH: psfSize, noiseFloor,
    });
    const bands = measureSpectrumBands(recovered, w, h);
    return { sigma, highBandEnergy: bands.highBandEnergy };
  });

  // 4. RACE: identity (no candidate filtered). FOLD: sequential Fibonacci
  //    walk + argmin |Binet residual| + sqrt(5)-Newton refinement.
  const { output } = await forkRaceFold<CandidateEnergy, SelectSigmaResult>({
    forks,
    fold: (energies): SelectSigmaResult => {
      let state: GoldenAttractorState = createGoldenAttractorState();
      const reports: SigmaCandidateReport[] = [];
      let bestIdx = -1;
      let bestAbsResidual = Infinity;

      for (let i = 0; i < energies.length; i++) {
        const { sigma, highBandEnergy } = energies[i]!;
        state = updateGoldenAttractor(state, highBandEnergy);
        const residual = binetResidual(state.deficitRatio);
        const excluded = state.cycle <= 1;
        reports.push({
          sigma,
          highBandEnergy,
          deficitRatio: state.deficitRatio,
          binetResidual: residual,
          excludedFromAttractor: excluded,
        });
        if (!excluded && Math.abs(residual) < bestAbsResidual) {
          bestAbsResidual = Math.abs(residual);
          bestIdx = i;
        }
      }

      // Fallback: if no eligible candidate (impossible with candidateCount>=3
      // since cycle increments to 2 by the second iteration), pick the largest
      // sigma so callers always get a usable result.
      if (bestIdx < 0) bestIdx = energies.length - 1;

      const selectedSigma = grid[bestIdx]!;
      let refinedSigma = selectedSigma;
      if (bestIdx >= 1) {
        refinedSigma = phiAttractorNewtonStep({
          sigma: grid[bestIdx]!,
          prevSigma: grid[bestIdx - 1]!,
          ratio: reports[bestIdx]!.deficitRatio,
          prevRatio: reports[bestIdx - 1]!.deficitRatio,
        });
      }

      return {
        selectedSigma,
        refinedSigma,
        candidates: reports,
        bestCandidateIndex: bestIdx,
      };
    },
  });

  return output;
}

/**
 * Synchronous public surface preserved for backward compatibility — the
 * Wiener forks are CPU-bound and (in single-thread Node) don't actually
 * benefit from being awaited individually, so the sync version just inlines
 * the same FRF shape without the Promise plumbing.
 */
export function selectSigmaByGoldenAttractor(args: SelectSigmaArgs): SelectSigmaResult {
  const {
    plane,
    w,
    h,
    psfSize = 11,
    sigmaMin = 0.5,
    candidateCount = 7,
    noiseFloor = 1e-3,
  } = args;

  if (!isPositivePow2(w)) {
    throw new RangeError(`selectSigmaByGoldenAttractor: w must be a positive power of 2, got ${w}`);
  }
  if (!isPositivePow2(h)) {
    throw new RangeError(`selectSigmaByGoldenAttractor: h must be a positive power of 2, got ${h}`);
  }
  if (plane.length !== w * h) {
    throw new RangeError(
      `selectSigmaByGoldenAttractor: plane.length=${plane.length} != w*h=${w * h}`,
    );
  }
  if (!Number.isInteger(psfSize) || psfSize <= 0 || (psfSize & 1) === 0) {
    throw new RangeError(
      `selectSigmaByGoldenAttractor: psfSize must be a positive odd integer, got ${psfSize}`,
    );
  }
  if (!Number.isFinite(sigmaMin) || sigmaMin <= 0) {
    throw new RangeError(`selectSigmaByGoldenAttractor: sigmaMin must be > 0, got ${sigmaMin}`);
  }
  if (!Number.isInteger(candidateCount) || candidateCount < 3) {
    throw new RangeError(
      `selectSigmaByGoldenAttractor: candidateCount must be an integer >= 3, got ${candidateCount}`,
    );
  }

  const grid = goldenSpacedGrid(sigmaMin, candidateCount);
  let state: GoldenAttractorState = createGoldenAttractorState();
  const reports: SigmaCandidateReport[] = [];
  let bestIdx = -1;
  let bestAbsResidual = Infinity;

  for (let i = 0; i < grid.length; i++) {
    const sigma = grid[i]!;
    const psf = gaussianPSF(sigma, psfSize);
    const recovered = wienerDeblur({
      plane, w, h, psf, psfW: psfSize, psfH: psfSize, noiseFloor,
    });
    const bands = measureSpectrumBands(recovered, w, h);
    state = updateGoldenAttractor(state, bands.highBandEnergy);
    const residual = binetResidual(state.deficitRatio);
    const excluded = state.cycle <= 1;
    reports.push({
      sigma,
      highBandEnergy: bands.highBandEnergy,
      deficitRatio: state.deficitRatio,
      binetResidual: residual,
      excludedFromAttractor: excluded,
    });
    if (!excluded && Math.abs(residual) < bestAbsResidual) {
      bestAbsResidual = Math.abs(residual);
      bestIdx = i;
    }
  }

  if (bestIdx < 0) bestIdx = grid.length - 1;

  const selectedSigma = grid[bestIdx]!;
  let refinedSigma = selectedSigma;
  if (bestIdx >= 1) {
    refinedSigma = phiAttractorNewtonStep({
      sigma: grid[bestIdx]!,
      prevSigma: grid[bestIdx - 1]!,
      ratio: reports[bestIdx]!.deficitRatio,
      prevRatio: reports[bestIdx - 1]!.deficitRatio,
    });
  }

  return {
    selectedSigma,
    refinedSigma,
    candidates: reports,
    bestCandidateIndex: bestIdx,
  };
}

// Re-export PHI for convenience of callers reasoning about the grid spacing.
export { PHI };
