// @a0n/.ignored_gnosis/distributed-inference-host/src/spectral-noise-monitor.ts
import { mapPolyglot, parseMonsterPath } from './universal-signal-map';
import type { MonsterPath, PolyglotOp } from './universal-signal-map';

export type SpectralNoiseColor = 'white' | 'pink' | 'brown';

export type SpectralNoiseRuntimeState =
  | 'normal_carrier'
  | 'pre_collapse'
  | 'boundary_collapse';

export type SpectralNoiseAction =
  | 'none'
  | 'recommend_reshard'
  | 'recommend_recovery';

export interface SpectralNoiseFingerprint {
  color: SpectralNoiseColor;
  lowBand: number;
  highBand: number;
  slopeMagnitude: 0 | 1 | 2;
}

export interface BuleyProjection {
  waste: number;
  opportunity: number;
  diversity: number;
  score: number;
}

export type BuleyFace = 'waste' | 'opportunity' | 'diversity';

const ATTENTION_TRITON_PHASE_COUNT = 3;
const ATTENTION_AEON_RESOLUTION = 12;
const ATTENTION_HEXON_COUPLING = 6;

export const OPTIMIZER_ALIAS_REQUIRES_RESOLUTION_LIFT_THEOREM =
  'Gnosis.UniversalIntelligenceSSMClosure.optimizer_alias_requires_resolution_lift';
export const OPTIMIZER_RESIDUE_REQUIRES_RESOLUTION_LIFT_THEOREM =
  'Gnosis.UniversalIntelligenceSSMClosure.optimizer_residue_requires_resolution_lift';
export const OPTIMIZER_READY_ADMITS_RUNTIME_WORK_THEOREM =
  'Gnosis.UniversalIntelligenceSSMClosure.optimizer_ready_admits_runtime_work';
export const AEON_EIGHT_HEAD_OPTIMIZER_REQUIRES_LIFT_THEOREM =
  'Gnosis.UniversalIntelligenceSSMClosure.aeon_eight_head_optimizer_requires_lift';
export const LIFTED_EIGHT_HEAD_OPTIMIZER_ADMITS_RUNTIME_WORK_THEOREM =
  'Gnosis.UniversalIntelligenceSSMClosure.lifted_eight_head_optimizer_admits_runtime_work';
export const LIFTED_PINK_EIGHT_HEAD_OPTIMIZER_ADMITS_RUNTIME_WORK_THEOREM =
  'Gnosis.UniversalIntelligenceSSMClosure.lifted_pink_eight_head_optimizer_admits_runtime_work';

export type MeshAttentionClosureTheorem =
  | typeof OPTIMIZER_ALIAS_REQUIRES_RESOLUTION_LIFT_THEOREM
  | typeof OPTIMIZER_RESIDUE_REQUIRES_RESOLUTION_LIFT_THEOREM
  | typeof OPTIMIZER_READY_ADMITS_RUNTIME_WORK_THEOREM
  | typeof AEON_EIGHT_HEAD_OPTIMIZER_REQUIRES_LIFT_THEOREM
  | typeof LIFTED_EIGHT_HEAD_OPTIMIZER_ADMITS_RUNTIME_WORK_THEOREM
  | typeof LIFTED_PINK_EIGHT_HEAD_OPTIMIZER_ADMITS_RUNTIME_WORK_THEOREM;

export const AEON_EIGHT_HEAD_OPTIMIZER_THEOREM_LINEAGE = [
  AEON_EIGHT_HEAD_OPTIMIZER_REQUIRES_LIFT_THEOREM,
  OPTIMIZER_ALIAS_REQUIRES_RESOLUTION_LIFT_THEOREM,
] as const satisfies readonly MeshAttentionClosureTheorem[];
export const ALIASED_OPTIMIZER_THEOREM_LINEAGE = [
  OPTIMIZER_ALIAS_REQUIRES_RESOLUTION_LIFT_THEOREM,
] as const satisfies readonly MeshAttentionClosureTheorem[];
export const RESIDUE_OPTIMIZER_THEOREM_LINEAGE = [
  OPTIMIZER_RESIDUE_REQUIRES_RESOLUTION_LIFT_THEOREM,
] as const satisfies readonly MeshAttentionClosureTheorem[];
export const LIFTED_EIGHT_HEAD_OPTIMIZER_THEOREM_LINEAGE = [
  LIFTED_EIGHT_HEAD_OPTIMIZER_ADMITS_RUNTIME_WORK_THEOREM,
  OPTIMIZER_READY_ADMITS_RUNTIME_WORK_THEOREM,
] as const satisfies readonly MeshAttentionClosureTheorem[];
export const LIFTED_PINK_EIGHT_HEAD_OPTIMIZER_THEOREM_LINEAGE = [
  LIFTED_PINK_EIGHT_HEAD_OPTIMIZER_ADMITS_RUNTIME_WORK_THEOREM,
  OPTIMIZER_READY_ADMITS_RUNTIME_WORK_THEOREM,
] as const satisfies readonly MeshAttentionClosureTheorem[];
export const READY_OPTIMIZER_THEOREM_LINEAGE = [
  OPTIMIZER_READY_ADMITS_RUNTIME_WORK_THEOREM,
] as const satisfies readonly MeshAttentionClosureTheorem[];
export const UNPROVED_THRESHOLD_POLICY_THEOREM_LINEAGE =
  [] as const satisfies readonly MeshAttentionClosureTheorem[];

export type AttentionClosureKind =
  | 'aliased_noise'
  | 'structured_signal'
  | 'separated_residue'
  | 'aliased_closed';

export type ResolutionLiftActionForMesh = 'none' | 'lift_resolution';

export interface AttentionClosureInput {
  heads?: number;
  coupling?: number;
  observerResolution?: number;
}

export interface AttentionClosureState {
  heads: number;
  coupling: number;
  observerResolution: number;
  sourceResolution: number;
  foldedCoordinate: number;
  unresolvedResidue: number;
  headTrace: number[];
  traceSeparated: boolean;
  kind: AttentionClosureKind;
}

export interface DarkDeceptaconResolutionLiftDecision {
  action: ResolutionLiftActionForMesh;
  before: AttentionClosureState;
  after: AttentionClosureState;
  reason: string;
}

export const vacuumBuleyProjection: BuleyProjection = {
  waste: 0,
  opportunity: 0,
  diversity: 0,
  score: 0,
};

/**
 * Handles the gnosis clinamen Lift workflow.
 */
export function clinamenLift(
  b: BuleyProjection,
  face: BuleyFace
): BuleyProjection {
  const next: BuleyProjection = { ...b };
  next[face] = b[face] + 1;
  next.score = next.waste + next.opportunity + next.diversity;
  return next;
}

/**
 * Handles the gnosis clinamen Contract workflow.
 */
export function clinamenContract(
  b: BuleyProjection,
  face: BuleyFace
): BuleyProjection {
  const next: BuleyProjection = { ...b };
  next[face] = Math.max(0, b[face] - 1);
  next.score = next.waste + next.opportunity + next.diversity;
  return next;
}

/**
 * Handles the gnosis waste Face workflow.
 */
export function wasteFace(b: BuleyProjection): BuleyProjection {
  return { waste: b.waste, opportunity: 0, diversity: 0, score: b.waste };
}

/**
 * Handles the gnosis action Face workflow.
 */
export function actionFace(b: BuleyProjection): BuleyProjection {
  return {
    waste: 0,
    opportunity: b.opportunity,
    diversity: 0,
    score: b.opportunity,
  };
}

/**
 * Handles the gnosis entropy Face workflow.
 */
export function entropyFace(b: BuleyProjection): BuleyProjection {
  return {
    waste: 0,
    opportunity: 0,
    diversity: b.diversity,
    score: b.diversity,
  };
}

/**
 * Handles the gnosis cycle Permute workflow.
 */
export function cyclePermute(b: BuleyProjection): BuleyProjection {
  return {
    waste: b.opportunity,
    opportunity: b.diversity,
    diversity: b.waste,
    score: b.score,
  };
}

/**
 * Self-Similarity Violation: a Bule unit's score that exceeds the
 * manifold's phase-count ceiling. Mirrors
 * `Gnosis.BuleySelfSimilarityViolation` in Lean.
 */
export interface SelfSimilarityViolation {
  ceiling: number;
  score: number;
  correctiveContractCount: number;
}

/**
 * Handles the gnosis self Similarity Violation workflow.
 */
export function selfSimilarityViolation(
  b: BuleyProjection,
  ceiling: number
): SelfSimilarityViolation | null {
  if (b.score <= ceiling) return null;
  return {
    ceiling,
    score: b.score,
    correctiveContractCount: b.score - ceiling,
  };
}

/**
 * Manifold ceilings for the canonical tower levels — Triton(3),
 * Hexon(6), Enneon(9), Trihexon(18), Trihexenneon(54). The coordinator
 * uses these to recognize "X-pressure in a Y-manifold" violations.
 */
export const TOWER_CEILINGS = {
  triton: 3,
  hexon: 6,
  enneon: 9,
  decagon: 10,
  hendecagon: 11,
  dodecagon: 12,
  trihexon: 18,
  bosonicString: 26,
  trihexenneon: 54,
} as const;
// Note: `decagon`, `hendecagon`, and `bosonicString` mirror the
// string-theory dimension signatures in
// `Gnosis.BraidedTower.{decagon_in_tower, hendecagon_in_tower,
// bosonic_string_in_tower}`. They are signatures, not derivations:
// the cost-algebra tower permits any phase count via
// `every_phase_count_is_a_tower`, and these constants name three
// specific walls that string theory happens to occupy. The runtime
// uses them only as ceiling defaults for governance reporting.

/**
 * Fingerprint pressure as a Bule projection. Mirrors
 * `Gnosis.TopologicalMetabolismBuleyBridge.fingerprintBuleyUnit`:
 * low-band mass maps to waste, high-band mass to opportunity, and spectral
 * slope magnitude maps to diversity.
 */
export function fingerprintBuleyProjection(
  fingerprint: SpectralNoiseFingerprint
): BuleyProjection {
  const waste = fingerprint.lowBand;
  const opportunity = fingerprint.highBand;
  const diversity = fingerprint.slopeMagnitude;
  return {
    waste,
    opportunity,
    diversity,
    score: waste + opportunity + diversity,
  };
}

/**
 * Governance-level self-similarity pressure. The default ceiling is Hexon(6),
 * matching the two-head attention ceiling proved in Lean.
 */
export function governanceSelfSimilarityViolation(
  fingerprint: SpectralNoiseFingerprint,
  ceiling: number = TOWER_CEILINGS.hexon
): SelfSimilarityViolation | null {
  return selfSimilarityViolation(
    fingerprintBuleyProjection(fingerprint),
    ceiling
  );
}

export interface MeshAttentionClosureConfig extends AttentionClosureInput {
  residueThreshold?: number;
  requestedSpeculativeDepth?: number;
}

export interface MeshAttentionClosureAssessment {
  observed: AttentionClosureState;
  decision: DarkDeceptaconResolutionLiftDecision;
  action: ResolutionLiftActionForMesh;
  workPolicy: MeshAttentionClosureWorkPolicy;
}

export interface MeshAttentionClosureWorkPolicy {
  preserveHeads: boolean;
  allowHeadPruning: boolean;
  allowSpeculativeDecode: boolean;
  allowStandingWaveCompression: boolean;
  routeAction: ResolutionLiftActionForMesh;
  speculativeDepthLimit: number;
  telemetryLabel:
    | 'attention_closure_lift_required'
    | 'attention_closure_structured';
  theoremLineage: readonly MeshAttentionClosureTheorem[];
}

export interface MeshAttentionClosureWorkload {
  hiddenDim: number;
  batchSize: number;
  layerCount: number;
  standingCount: number;
  requestedSpeculativeDepth: number;
  speculativeCandidates: number;
  headPruneCandidates: number;
  headDim: number;
}

export interface MeshAttentionClosureWorkEstimate {
  baselineOptimizerUnits: number;
  gatedOptimizerUnits: number;
  avoidedOptimizerUnits: number;
  avoidedCompressDecompressUnits: number;
  avoidedSpeculativeUnits: number;
  preservedHeadPruneUnits: number;
  fullRouteBytesWhenLifted: number;
  compressedRouteBytesIfStable: number;
}

function natural(value: number): number {
  if (!Number.isFinite(value)) return 0;
  return Math.max(0, Math.trunc(value));
}

function positiveNatural(value: number, fallback: number): number {
  const n = natural(value);
  return n > 0 ? n : fallback;
}

function mod(n: number, modulus: number): number {
  const m = positiveNatural(modulus, 1);
  return ((natural(n) % m) + m) % m;
}

function attentionClosureKind(
  traceSeparated: boolean,
  unresolvedResidue: number
): AttentionClosureKind {
  if (traceSeparated && unresolvedResidue === 0) return 'structured_signal';
  if (!traceSeparated && unresolvedResidue > 0) return 'aliased_noise';
  if (traceSeparated && unresolvedResidue > 0) return 'separated_residue';
  return 'aliased_closed';
}

function createMeshAttentionClosureState(
  input: AttentionClosureInput = {}
): AttentionClosureState {
  const heads = natural(input.heads ?? 8);
  const coupling = natural(input.coupling ?? 0);
  const sourceResolution = ATTENTION_TRITON_PHASE_COUNT * heads + coupling;
  const observerResolution = positiveNatural(
    input.observerResolution ?? ATTENTION_AEON_RESOLUTION,
    ATTENTION_AEON_RESOLUTION
  );
  const foldedCoordinate = mod(sourceResolution, observerResolution);
  const unresolvedResidue = Math.max(0, sourceResolution - observerResolution);
  const headTrace = Array.from({ length: heads }, (_, head) =>
    mod(ATTENTION_TRITON_PHASE_COUNT * (head + 1), observerResolution)
  );
  const traceSeparated = new Set(headTrace).size === headTrace.length;

  return {
    heads,
    coupling,
    observerResolution,
    sourceResolution,
    foldedCoordinate,
    unresolvedResidue,
    headTrace,
    traceSeparated,
    kind: attentionClosureKind(traceSeparated, unresolvedResidue),
  };
}

function applyMeshDarkDeceptaconResolutionLift(
  state: AttentionClosureState,
  residueThreshold = 1
): DarkDeceptaconResolutionLiftDecision {
  const threshold = Math.max(1, natural(residueThreshold));
  const shouldLift =
    !state.traceSeparated || state.unresolvedResidue >= threshold;
  if (!shouldLift) {
    return {
      action: 'none',
      before: state,
      after: state,
      reason:
        'attention trace already separated under the current observer resolution',
    };
  }

  return {
    action: 'lift_resolution',
    before: state,
    after: createMeshAttentionClosureState({
      heads: state.heads,
      coupling: state.coupling,
      observerResolution: state.sourceResolution,
    }),
    reason:
      'attention trace is aliased or leaking; lift observer resolution to the source phase',
  };
}

function isMeshAeonEightHeadAliasWitness(
  state: AttentionClosureState
): boolean {
  return (
    state.heads === 8 &&
    state.coupling === 0 &&
    state.observerResolution === ATTENTION_AEON_RESOLUTION &&
    state.sourceResolution === ATTENTION_TRITON_PHASE_COUNT * 8
  );
}

function isMeshLiftedEightHeadClosureWitness(
  state: AttentionClosureState
): boolean {
  return (
    state.heads === 8 &&
    state.coupling === 0 &&
    state.observerResolution === ATTENTION_TRITON_PHASE_COUNT * 8 &&
    state.sourceResolution === ATTENTION_TRITON_PHASE_COUNT * 8 &&
    state.traceSeparated &&
    state.unresolvedResidue === 0
  );
}

function isMeshLiftedPinkEightHeadClosureWitness(
  state: AttentionClosureState
): boolean {
  const sourceResolution =
    ATTENTION_TRITON_PHASE_COUNT * 8 + ATTENTION_HEXON_COUPLING;
  return (
    state.heads === 8 &&
    state.coupling === ATTENTION_HEXON_COUPLING &&
    state.observerResolution === sourceResolution &&
    state.sourceResolution === sourceResolution &&
    state.traceSeparated &&
    state.unresolvedResidue === 0
  );
}

/**
 * Handles the gnosis mesh Attention Closure Theorem Lineage workflow.
 */
export function meshAttentionClosureTheoremLineage(
  decision: DarkDeceptaconResolutionLiftDecision
): readonly MeshAttentionClosureTheorem[] {
  if (decision.action === 'lift_resolution') {
    if (isMeshAeonEightHeadAliasWitness(decision.before)) {
      return AEON_EIGHT_HEAD_OPTIMIZER_THEOREM_LINEAGE;
    }
    if (!decision.before.traceSeparated) {
      return ALIASED_OPTIMIZER_THEOREM_LINEAGE;
    }
    return RESIDUE_OPTIMIZER_THEOREM_LINEAGE;
  }

  if (isMeshLiftedPinkEightHeadClosureWitness(decision.after)) {
    return LIFTED_PINK_EIGHT_HEAD_OPTIMIZER_THEOREM_LINEAGE;
  }
  if (isMeshLiftedEightHeadClosureWitness(decision.after)) {
    return LIFTED_EIGHT_HEAD_OPTIMIZER_THEOREM_LINEAGE;
  }
  if (decision.after.traceSeparated && decision.after.unresolvedResidue === 0) {
    return READY_OPTIMIZER_THEOREM_LINEAGE;
  }
  return UNPROVED_THRESHOLD_POLICY_THEOREM_LINEAGE;
}

/**
 * Handles the gnosis mesh Attention Closure Work Policy workflow.
 */
export function meshAttentionClosureWorkPolicy(
  decision: DarkDeceptaconResolutionLiftDecision,
  requestedSpeculativeDepth = 1
): MeshAttentionClosureWorkPolicy {
  const liftRequired = decision.action === 'lift_resolution';
  return {
    preserveHeads: liftRequired,
    allowHeadPruning: !liftRequired,
    allowSpeculativeDecode: !liftRequired,
    allowStandingWaveCompression: !liftRequired,
    routeAction: liftRequired ? 'lift_resolution' : 'none',
    speculativeDepthLimit: liftRequired
      ? 1
      : Math.max(1, natural(requestedSpeculativeDepth)),
    telemetryLabel: liftRequired
      ? 'attention_closure_lift_required'
      : 'attention_closure_structured',
    theoremLineage: meshAttentionClosureTheoremLineage(decision),
  };
}

/**
 * Handles the gnosis estimate Mesh Attention Closure Saved Work workflow.
 */
export function estimateMeshAttentionClosureSavedWork(
  policy: MeshAttentionClosureWorkPolicy,
  workload: MeshAttentionClosureWorkload
): MeshAttentionClosureWorkEstimate {
  const hiddenDim = natural(workload.hiddenDim);
  const batchSize = Math.max(1, natural(workload.batchSize));
  const layerCount = Math.max(1, natural(workload.layerCount));
  const standingCount = Math.min(natural(workload.standingCount), hiddenDim);
  const requestedDepth = Math.max(
    1,
    natural(workload.requestedSpeculativeDepth)
  );
  const compressDecompressUnits =
    layerCount * batchSize * (standingCount + hiddenDim);
  const speculativeUnits =
    natural(workload.speculativeCandidates) * requestedDepth;
  const headPruneUnits =
    natural(workload.headPruneCandidates) *
    Math.max(1, natural(workload.headDim));
  const baselineOptimizerUnits =
    compressDecompressUnits + speculativeUnits + headPruneUnits;
  const gatedCompressDecompressUnits = policy.allowStandingWaveCompression
    ? compressDecompressUnits
    : 0;
  const gatedSpeculativeUnits = policy.allowSpeculativeDecode
    ? natural(workload.speculativeCandidates) *
      Math.max(1, policy.speculativeDepthLimit)
    : 0;
  const gatedHeadPruneUnits = policy.allowHeadPruning ? headPruneUnits : 0;
  const gatedOptimizerUnits =
    gatedCompressDecompressUnits + gatedSpeculativeUnits + gatedHeadPruneUnits;

  return {
    baselineOptimizerUnits,
    gatedOptimizerUnits,
    avoidedOptimizerUnits: Math.max(
      0,
      baselineOptimizerUnits - gatedOptimizerUnits
    ),
    avoidedCompressDecompressUnits: Math.max(
      0,
      compressDecompressUnits - gatedCompressDecompressUnits
    ),
    avoidedSpeculativeUnits: Math.max(
      0,
      speculativeUnits - gatedSpeculativeUnits
    ),
    preservedHeadPruneUnits: Math.max(0, headPruneUnits - gatedHeadPruneUnits),
    fullRouteBytesWhenLifted: layerCount * batchSize * hiddenDim * 4,
    compressedRouteBytesIfStable: layerCount * batchSize * standingCount * 4,
  };
}

/**
 * Handles the gnosis classify Mesh Attention Closure workflow.
 */
export function classifyMeshAttentionClosure(
  config: MeshAttentionClosureConfig = {}
): MeshAttentionClosureAssessment {
  const observed = createMeshAttentionClosureState(config);
  const decision = applyMeshDarkDeceptaconResolutionLift(
    observed,
    config.residueThreshold
  );
  return {
    observed,
    decision,
    action: decision.action,
    workPolicy: meshAttentionClosureWorkPolicy(
      decision,
      config.requestedSpeculativeDepth
    ),
  };
}

export interface LayerNodeTrace {
  requestId: string;
  nodeIdx: number;
  chunkIdx: number;
  startPos: number;
  sequenceLength: number;
  latencyMs: number;
  expectedLatencyMs: number;
  requestEntropy: number;
  boundaryLimit: number;
  timestamp: number;
}

export interface SpectralNoiseAssessment {
  state: SpectralNoiseRuntimeState;
  action: SpectralNoiseAction;
  fingerprint: SpectralNoiseFingerprint;
  fingerprintBuley: BuleyProjection;
  governanceViolation: SelfSimilarityViolation | null;
  attentionClosure?: MeshAttentionClosureAssessment;
  buley: BuleyProjection;
  trace: LayerNodeTrace;
  reason: string;
  monsterPath?: MonsterPath;
}

export interface SpectralNoiseMonitorConfig {
  expectedLatencyMs?: number;
  boundaryLimit?: number;
  preCollapseWindow?: number;
  buleyCeiling?: number;
  attentionClosure?: MeshAttentionClosureConfig;
  log?: (assessment: SpectralNoiseAssessment) => void;
}

const DEFAULT_EXPECTED_LATENCY_MS = 25;
const DEFAULT_BOUNDARY_LIMIT = 8;
const DEFAULT_PRE_COLLAPSE_WINDOW = 3;

/**
 * Handles the gnosis request Entropy workflow.
 */
export function requestEntropy(
  sequenceLength: number,
  nodeIdx: number
): number {
  return Math.max(
    0,
    Math.ceil(Math.log2(Math.max(2, sequenceLength + nodeIdx + 1)))
  );
}

/**
 * Handles the gnosis reduce Trace Fingerprint workflow.
 */
export function reduceTraceFingerprint(
  trace: LayerNodeTrace
): SpectralNoiseFingerprint {
  const latencyRatio =
    trace.expectedLatencyMs > 0 ? trace.latencyMs / trace.expectedLatencyMs : 0;

  if (latencyRatio >= 2) {
    return { color: 'brown', lowBand: 4, highBand: 1, slopeMagnitude: 2 };
  }
  if (latencyRatio >= 1) {
    return { color: 'pink', lowBand: 2, highBand: 1, slopeMagnitude: 1 };
  }
  return { color: 'white', lowBand: 1, highBand: 1, slopeMagnitude: 0 };
}

/**
 * Handles the gnosis measure Buley workflow.
 */
export function measureBuley(
  trace: LayerNodeTrace,
  fingerprint: SpectralNoiseFingerprint,
  recentFingerprints: readonly SpectralNoiseFingerprint[] = []
): BuleyProjection {
  const waste = Math.max(0, trace.requestEntropy - trace.boundaryLimit);
  const opportunity = Math.max(0, trace.boundaryLimit - trace.requestEntropy);
  const diversity = new Set([
    fingerprint.color,
    ...recentFingerprints.map((fp) => fp.color),
  ]).size;
  return {
    waste,
    opportunity,
    diversity,
    score: waste + opportunity + diversity,
  };
}

/**
 * Handles the gnosis classify Spectral Noise workflow.
 */
export function classifySpectralNoise(
  trace: LayerNodeTrace,
  recentFingerprints: readonly SpectralNoiseFingerprint[] = [],
  buleyCeiling: number = TOWER_CEILINGS.hexon
): SpectralNoiseAssessment {
  const fingerprint = reduceTraceFingerprint(trace);
  const fingerprintBuley = fingerprintBuleyProjection(fingerprint);
  const governanceViolation = selfSimilarityViolation(
    fingerprintBuley,
    buleyCeiling
  );
  const buley = measureBuley(trace, fingerprint, recentFingerprints);
  const boundaryCollapsed = trace.requestEntropy > trace.boundaryLimit;
  if (boundaryCollapsed) {
    return {
      state: 'boundary_collapse',
      action: 'recommend_recovery',
      fingerprint,
      fingerprintBuley,
      governanceViolation,
      buley,
      trace,
      reason: 'request entropy exceeded the boundary limit',
    };
  }

  const saturatedCount = recentFingerprints.filter(
    (fp) => fp.slopeMagnitude === 2
  ).length;
  if (
    fingerprint.slopeMagnitude === 2 &&
    recentFingerprints.length > 0 &&
    saturatedCount === recentFingerprints.length
  ) {
    return {
      state: 'pre_collapse',
      action: 'recommend_reshard',
      fingerprint,
      fingerprintBuley,
      governanceViolation,
      buley,
      trace,
      reason: 'magnitude-two fingerprint repeated below the entropy boundary',
    };
  }

  return {
    state: 'normal_carrier',
    action: 'none',
    fingerprint,
    fingerprintBuley,
    governanceViolation,
    buley,
    trace,
    reason: 'carrier remains inside spectral and entropy contracts',
  };
}

export class SpectralNoiseMonitor {
  private readonly expectedLatencyMs: number;
  private readonly boundaryLimit: number;
  private readonly preCollapseWindow: number;
  private readonly buleyCeiling: number;
  private readonly attentionClosureConfig:
    | MeshAttentionClosureConfig
    | undefined;
  private readonly log:
    | ((assessment: SpectralNoiseAssessment) => void)
    | undefined;
  private readonly fingerprints: SpectralNoiseFingerprint[] = [];
  private readonly assessments: SpectralNoiseAssessment[] = [];

  constructor(config: SpectralNoiseMonitorConfig = {}) {
    this.expectedLatencyMs =
      config.expectedLatencyMs ?? DEFAULT_EXPECTED_LATENCY_MS;
    this.boundaryLimit = config.boundaryLimit ?? DEFAULT_BOUNDARY_LIMIT;
    this.preCollapseWindow =
      config.preCollapseWindow ?? DEFAULT_PRE_COLLAPSE_WINDOW;
    this.buleyCeiling = config.buleyCeiling ?? TOWER_CEILINGS.hexon;
    this.attentionClosureConfig = config.attentionClosure;
    this.log = config.log;
  }

  ingest(
    trace: Omit<LayerNodeTrace, 'expectedLatencyMs' | 'boundaryLimit'> & {
      expectedLatencyMs?: number;
      boundaryLimit?: number;
    }
  ): SpectralNoiseAssessment {
    const normalized: LayerNodeTrace = {
      ...trace,
      expectedLatencyMs: trace.expectedLatencyMs ?? this.expectedLatencyMs,
      boundaryLimit: trace.boundaryLimit ?? this.boundaryLimit,
    };

    const recent = this.fingerprints.slice(
      Math.max(0, this.fingerprints.length - (this.preCollapseWindow - 1))
    );
    const assessment = classifySpectralNoise(
      normalized,
      recent,
      this.buleyCeiling
    );
    if (this.attentionClosureConfig !== undefined) {
      assessment.attentionClosure = classifyMeshAttentionClosure(
        this.attentionClosureConfig
      );
    }
    this.fingerprints.push(assessment.fingerprint);
    this.assessments.push(assessment);

    if (assessment.action !== 'none') {
      this.log?.(assessment);
    }

    return assessment;
  }

  /**
   * Ingest a Polyglot IR operation, mapping it to its deterministic
   * Monster Path and updating the monitor with its topological cost.
   */
  ingestOp(
    op: PolyglotOp,
    trace: Omit<
      LayerNodeTrace,
      'expectedLatencyMs' | 'boundaryLimit' | 'requestEntropy'
    > & {
      expectedLatencyMs?: number;
      boundaryLimit?: number;
    }
  ): SpectralNoiseAssessment {
    const path = mapPolyglot(op);
    const entropy = path.length;
    const assessment = this.ingest({
      ...trace,
      requestEntropy: entropy,
    });
    assessment.monsterPath = path;
    return assessment;
  }

  /**
   * Ingest a raw topological fingerprint (Monster Path), updating the monitor
   * with its topological cost.
   */
  ingestFingerprint(
    fingerprint: string | MonsterPath,
    trace: Omit<
      LayerNodeTrace,
      'expectedLatencyMs' | 'boundaryLimit' | 'requestEntropy'
    > & {
      expectedLatencyMs?: number;
      boundaryLimit?: number;
    }
  ): SpectralNoiseAssessment {
    const path =
      typeof fingerprint === 'string'
        ? parseMonsterPath(fingerprint)
        : fingerprint;
    const entropy = path.length;
    const assessment = this.ingest({
      ...trace,
      requestEntropy: entropy,
    });
    assessment.monsterPath = path;
    return assessment;
  }

  getAssessments(): SpectralNoiseAssessment[] {
    return this.assessments.slice();
  }
}
