// @a0n/.ignored_gnosis/distributed-inference-host/src/stateless-mesh-projection.ts
/**
 * Stateless mesh projection helpers.
 *
 * The renderer owns only reusable UV/index buffers. Per-image shape stays in
 * depth/alpha texture planes produced by the sieve/depth pipeline.
 */

export const STATELESS_MESH_PROJECTION_SCHEMA_VERSION = 1 as const;

export interface StatelessMeshGridOptions {
  readonly width: number;
  readonly height: number;
  /**
   * Include a back plane and perimeter wall index ranges. Use this when the
   * caller wants a solid extrusion rather than a front heightfield shell.
   */
  readonly solid?: boolean;
}

export interface StatelessMeshGridBuffers {
  readonly schemaVersion: typeof STATELESS_MESH_PROJECTION_SCHEMA_VERSION;
  readonly width: number;
  readonly height: number;
  /** UV coordinates, row-major: `u, v` per vertex. Front vertices come first. */
  readonly uv: Float32Array;
  /** Front heightfield triangles, and optional back plane triangles. */
  readonly surfaceIndices: Uint32Array;
  /**
   * Static outer rectangle walls. Alpha-derived interior walls require
   * `buildAlphaBoundaryWallIndices`.
   */
  readonly perimeterWallIndices: Uint32Array;
  readonly vertexCount: number;
}

export interface AlphaBoundaryOptions {
  readonly width: number;
  readonly height: number;
  readonly alpha: Float32Array | Uint8Array;
  /** Pixels with alpha >= threshold are admitted. Default: `0.5`. */
  readonly threshold?: number;
}

export interface ProjectionTextureContract {
  readonly width: number;
  readonly height: number;
  readonly depth: Float32Array;
  readonly alpha: Float32Array | Uint8Array;
  readonly confidence?: Float32Array;
}

export interface ProjectionTextureValidation {
  readonly ok: boolean;
  readonly pixelCount: number;
  readonly errors: readonly string[];
}

export const STATELESS_EXTRUSION_WGSL = `
struct VertexIn {
  @location(0) uv: vec2<f32>,
};

struct VertexOut {
  @builtin(position) position: vec4<f32>,
  @location(0) uv: vec2<f32>,
  @location(1) admission: f32,
};

@group(0) @binding(0) var depthTex: texture_2d<f32>;
@group(0) @binding(1) var alphaTex: texture_2d<f32>;
@group(0) @binding(2) var samp: sampler;

struct ExtrusionUniforms {
  depthScale: f32,
  alphaThreshold: f32,
  backPlane: f32,
  _pad: f32,
};

@group(0) @binding(3) var<uniform> uniforms: ExtrusionUniforms;

@vertex
fn vs_main(in: VertexIn) -> VertexOut {
  let depth = textureSampleLevel(depthTex, samp, in.uv, 0.0).r;
  let alpha = textureSampleLevel(alphaTex, samp, in.uv, 0.0).r;
  let xy = in.uv * vec2<f32>(2.0, -2.0) + vec2<f32>(-1.0, 1.0);
  var out: VertexOut;
  out.position = vec4<f32>(xy, depth * uniforms.depthScale, 1.0);
  out.uv = in.uv;
  out.admission = alpha;
  return out;
}

@fragment
fn fs_main(in: VertexOut) -> @location(0) vec4<f32> {
  if (in.admission < uniforms.alphaThreshold) {
    discard;
  }
  return vec4<f32>(in.uv, in.admission, 1.0);
}
`;

/**
 * Builds the Stateless Mesh Grid.
 */
export function buildStatelessMeshGrid(options: StatelessMeshGridOptions): StatelessMeshGridBuffers {
  const { width, height, solid = false } = options;
  assertGridShape(width, height);

  const planeVertexCount = width * height;
  const vertexCount = solid ? planeVertexCount * 2 : planeVertexCount;
  const uv = new Float32Array(vertexCount * 2);
  writeUvPlane(uv, 0, width, height);
  if (solid) {
    writeUvPlane(uv, planeVertexCount * 2, width, height);
  }

  const front = buildPlaneIndices(width, height, false, 0);
  const back = solid ? buildPlaneIndices(width, height, true, planeVertexCount) : new Uint32Array(0);
  const surfaceIndices = concatUint32(front, back);
  const perimeterWallIndices = solid ? buildPerimeterWallIndices(width, height, planeVertexCount) : new Uint32Array(0);

  return {
    schemaVersion: STATELESS_MESH_PROJECTION_SCHEMA_VERSION,
    width,
    height,
    uv,
    surfaceIndices,
    perimeterWallIndices,
    vertexCount,
  };
}

/**
 * Builds the Alpha Boundary Wall Indices.
 */
export function buildAlphaBoundaryWallIndices(options: AlphaBoundaryOptions): Uint32Array {
  const { width, height, alpha, threshold = 0.5 } = options;
  assertGridShape(width, height);
  const n = width * height;
  if (alpha.length !== n) {
    throw new Error(`alpha length ${alpha.length} does not match ${n}`);
  }

  const backOffset = n;
  const out: number[] = [];
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      const p = y * width + x;
      if (!admitted(alpha[p]!, threshold)) continue;
      if (x === 0 || !admitted(alpha[p - 1]!, threshold)) {
        pushWall(out, p, p, backOffset);
      }
      if (x + 1 === width || !admitted(alpha[p + 1]!, threshold)) {
        pushWall(out, p, p, backOffset);
      }
      if (y === 0 || !admitted(alpha[p - width]!, threshold)) {
        pushWall(out, p, p, backOffset);
      }
      if (y + 1 === height || !admitted(alpha[p + width]!, threshold)) {
        pushWall(out, p, p, backOffset);
      }
    }
  }
  return new Uint32Array(out);
}

/**
 * Handles the gnosis validate Projection Texture Contract workflow.
 */
export function validateProjectionTextureContract(contract: ProjectionTextureContract): ProjectionTextureValidation {
  const { width, height, depth, alpha, confidence } = contract;
  const errors: string[] = [];
  if (!Number.isInteger(width) || width <= 0) errors.push('width must be a positive integer');
  if (!Number.isInteger(height) || height <= 0) errors.push('height must be a positive integer');
  const pixelCount = Math.max(0, width * height);
  if (depth.length !== pixelCount) errors.push(`depth length ${depth.length} does not match ${pixelCount}`);
  if (alpha.length !== pixelCount) errors.push(`alpha length ${alpha.length} does not match ${pixelCount}`);
  if (confidence !== undefined && confidence.length !== pixelCount) {
    errors.push(`confidence length ${confidence.length} does not match ${pixelCount}`);
  }
  return { ok: errors.length === 0, pixelCount, errors };
}

function writeUvPlane(out: Float32Array, offset: number, width: number, height: number): void {
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      const i = offset + (y * width + x) * 2;
      out[i] = width === 1 ? 0.5 : x / (width - 1);
      out[i + 1] = height === 1 ? 0.5 : y / (height - 1);
    }
  }
}

function buildPlaneIndices(width: number, height: number, reverse: boolean, offset: number): Uint32Array {
  const indices = new Uint32Array((width - 1) * (height - 1) * 6);
  let k = 0;
  for (let y = 0; y < height - 1; y++) {
    for (let x = 0; x < width - 1; x++) {
      const i0 = offset + y * width + x;
      const i1 = i0 + 1;
      const i2 = i0 + width;
      const i3 = i2 + 1;
      if (reverse) {
        indices.set([i0, i3, i1, i0, i2, i3], k);
      } else {
        indices.set([i0, i1, i3, i0, i3, i2], k);
      }
      k += 6;
    }
  }
  return indices;
}

function buildPerimeterWallIndices(width: number, height: number, backOffset: number): Uint32Array {
  const out: number[] = [];
  for (let x = 0; x < width - 1; x++) {
    pushWall(out, x, x + 1, backOffset);
    const bottom = (height - 1) * width + x;
    pushWall(out, bottom + 1, bottom, backOffset);
  }
  for (let y = 0; y < height - 1; y++) {
    const left = y * width;
    pushWall(out, left + width, left, backOffset);
    const right = y * width + width - 1;
    pushWall(out, right, right + width, backOffset);
  }
  return new Uint32Array(out);
}

function pushWall(out: number[], a: number, b: number, backOffset: number): void {
  const af = a;
  const bf = b;
  const ab = a + backOffset;
  const bb = b + backOffset;
  out.push(af, bf, bb, af, bb, ab);
}

function concatUint32(a: Uint32Array, b: Uint32Array): Uint32Array {
  if (b.length === 0) return a;
  const out = new Uint32Array(a.length + b.length);
  out.set(a, 0);
  out.set(b, a.length);
  return out;
}

function admitted(value: number, threshold: number): boolean {
  const unit = value > 1 ? value / 255 : value;
  return unit >= threshold;
}

function assertGridShape(width: number, height: number): void {
  if (!Number.isInteger(width) || !Number.isInteger(height) || width < 2 || height < 2) {
    throw new Error('stateless mesh grid requires integer width and height >= 2');
  }
}
