// @a0n/.ignored_gnosis/distributed-inference-host/src/station-ws.ts
/**
 * WSStation — IStation backed by a persistent WebSocket /flow socket.
 *
 * Wave-18 substrate (Death #3): a 3-worker bench measured 75 % per-hop
 * mean save and 89 % p95 save vs HTTPS by reusing one warm WebSocket per
 * worker instead of one fetch() per token. See:
 *   docs/death3-ws-substrate-3worker-result.md
 *
 * What this class is
 * ------------------
 * A drop-in IStation that mirrors `bench-trisplit.ts` `FlowClient` (lines
 * 73-108) so the production Pipeline dispatcher can fan tokens through the
 * `/flow` route on `tri-g4-{a,g,d}-NN` workers without changing
 * pipeline.ts.
 *
 * Scope (intentionally narrow for wave-19):
 *   - `forward()`  — sends ONE aeon-flow PHASE_{A,B,C} frame on a
 *     persistent socket. This is the hot path the wave-18 bench measured.
 *   - All other IStation methods (`embed`, `lmHead`, `cobordism*`,
 *     `bizarroForward`, KV-session methods, decode-next, sample-next)
 *     delegate to an inner HTTPS `Station`. The /flow route on the
 *     deployed worker (apps/distributed-inference-worker/src/index.ts
 *     line 2254 `handleFlowUpgrade`) only knows split-a/b/c phases — it
 *     does NOT yet handle embed / lm-head / cobordism. Those once-per-
 *     token calls stay HTTPS by design.
 *
 * Why "wraps an HTTPS Station" instead of throw-on-unsupported:
 *   - Operators can swap WSStation in for any HTTPS Station without
 *     touching the cobordism fan-out, embed, or lm-head dispatch paths.
 *   - The hot path (per-hop forward) runs on the WS substrate; cold paths
 *     keep working on HTTPS. The 75 % save is on the per-hop count, which
 *     dominates the chain depth × token count.
 *   - When the worker /flow route is extended to cover embed/lm-head, we
 *     flip the delegation to WS one method at a time without breaking
 *     callers.
 *
 * Observability
 * -------------
 * Every `forward()` call emits a structured log line via `console.log`
 * when `WS_STATION_TRACE=1`:
 *   `[WSStation:forward] url=<base> phase=<A|B|C> layer=<n> seq=<n>
 *      bytes_in=<n> bytes_out=<n> ms=<n>`
 *
 * The HTTPS `Station` does NOT log per-call latency by default; flipping
 * `STATION_HTTPS_TRACE=1` opts it into the same log line via the wrapper
 * field below. Both modes share the same log shape so the 75 % per-hop
 * delta is visible in production by grepping a single log line.
 *
 * Config — staged rollout
 * ----------------------
 * The transport is selected at Station-construction time. Default is
 * HTTPS (the existing `Station`). Operators flip to WS by either:
 *
 *   1. Passing `transport: 'ws'` in `WSStationConfig` (programmatic).
 *   2. Setting `STATION_TRANSPORT=flow` in the Worker env — read by
 *      callers that opt-in (the host shim does NOT auto-read env to
 *      preserve back-compat for embedded callers).
 *
 * The default of HTTPS lets operators roll WS to one mesh, observe the
 * per-hop delta in logs, and expand only when stable.
 */

import {
  PHASE_A,
  PHASE_B,
  PHASE_C,
  PHASE_MASK,
  STREAM_SPLIT_A_BASE,
  STREAM_SPLIT_B_BASE,
  STREAM_SPLIT_C_BASE,
  FlowCodec,
  FlowSendBatcher,
  encodeInferencePayload,
} from '@a0n/aeon/flow';

import {
  AMPLITUHEDRON_CONTENT_TYPE,
  encodeCaptureRequest,
  decodeReplayResponse,
} from './amplituhedron-wire';
import { Station, type StationConfig } from './station';
import type { IStation } from './types';

// ── ambient WebSocket type shim ────────────────────────────────────────
// We rely on `globalThis.WebSocket` (Node 22+ has it built-in; CF Workers
// have it; browsers have it). We do NOT depend on @types/node WebSocket
// to keep this file portable across runtimes.
declare const WebSocket: {
  new (url: string, opts?: unknown): WebSocket;
};

const TRACE =
  typeof process !== 'undefined' && process.env?.['WS_STATION_TRACE'] === '1';

function trace(line: string): void {
  if (!TRACE) return;
  try {
    console.log(line);
  } catch {
    // ignore
  }
}

function encodeResidualPayload(startPos: number, sequenceLength: number, body: Uint8Array): Uint8Array {
  const payload = new Uint8Array(8 + body.byteLength);
  const view = new DataView(payload.buffer);
  view.setUint32(0, startPos >>> 0, true);
  view.setUint32(4, sequenceLength >>> 0, true);
  payload.set(body, 8);
  return encodeInferencePayload(payload);
}

/**
 * Read `GNOSIS_FRAME_COALESCE` env var. Default ON per the "full send"
 * directive — concurrent same-turn `forward()` calls coalesce into one
 * underlying `socket.send()` via `FlowSendBatcher` (microtask flush).
 *
 * Disable via `GNOSIS_FRAME_COALESCE=0` or `GNOSIS_FRAME_COALESCE=false`
 * (case-insensitive). Read once per WSStation construction so test
 * harnesses can flip the flag between stations.
 */
function readCoalesceFlag(): boolean {
  if (typeof process === 'undefined') return true;
  const raw = process.env?.['GNOSIS_FRAME_COALESCE'];
  if (raw === undefined) return true;
  if (raw === '0') return false;
  if (raw.toLowerCase() === 'false') return false;
  return true;
}

/** Phase the WS station fires on every `forward()` call. */
export type WSStationPhase = 'split-a' | 'split-b' | 'split-c';

const PHASE_FOR: Record<WSStationPhase, number> = {
  'split-a': PHASE_A,
  'split-b': PHASE_B,
  'split-c': PHASE_C,
};

const STREAM_BASE_FOR: Record<WSStationPhase, number> = {
  'split-a': STREAM_SPLIT_A_BASE,
  'split-b': STREAM_SPLIT_B_BASE,
  'split-c': STREAM_SPLIT_C_BASE,
};

export interface WSStationConfig extends StationConfig {
  /**
   * Trisplit phase this station owns. Maps to the PHASE_{A,B,C} flag and
   * the STREAM_SPLIT_{A,B,C}_BASE stream-id namespace on the wire. The
   * /flow handler in the deployed worker dispatches on
   * `(streamId namespace, phase)` to pick the kernel op.
   *
   * Required when `transport: 'ws'`. Ignored otherwise.
   */
  phase?: WSStationPhase;
  /**
   * Optional WebSocket connect timeout (ms). Default 10 000.
   */
  connectTimeoutMs?: number;
  /**
   * Optional Bearer token sent in the WS upgrade. Default `mesh-bench`,
   * matches the deployed `tri-g4-*` workers' `BEARER_TOKEN`.
   */
  bearerToken?: string;
}

interface PendingCall {
  resolve: (data: Float32Array) => void;
  reject: (err: Error) => void;
}

/**
 * WebSocket-backed inference station. One persistent socket per WSStation
 * instance; the dispatcher reuses it for every `forward()` call.
 */
export class WSStation implements IStation {
  readonly startLayer: number;
  readonly endLayer: number;
  readonly role: 'entry' | 'mid' | 'exit' | 'cobordism' | undefined;
  readonly maxBatchSize: number;
  readonly phase: WSStationPhase;

  private readonly url: string;
  private readonly bearerToken: string;
  private readonly connectTimeoutMs: number;
  private readonly httpsFallback: Station;
  private readonly pending = new Map<number, PendingCall>();
  private readonly coalesceEnabled: boolean;
  private nextSequence = 1;
  private socket: WebSocket | null = null;
  private codec: FlowCodec | null = null;
  private connectPromise: Promise<void> | null = null;
  // One batcher per WSStation instance — bound to this station's socket so
  // cross-station coalescing never happens (each socket has its own write).
  private sendBatcher: FlowSendBatcher | null = null;

  constructor(config: WSStationConfig) {
    if (!config.phase) {
      throw new Error(
        `WSStation: 'phase' is required (one of split-a, split-b, split-c)`,
      );
    }
    if (typeof config.endpoint !== 'string') {
      throw new Error(
        `WSStation: 'endpoint' must be a URL string (got non-string endpoint)`,
      );
    }
    if (!Number.isInteger(config.startLayer) || config.startLayer < 0) {
      throw new Error(
        `WSStation: startLayer must be a non-negative integer (got ${String(config.startLayer)})`,
      );
    }
    if (!Number.isInteger(config.endLayer) || config.endLayer <= config.startLayer) {
      throw new Error(
        `WSStation: endLayer must be > startLayer (got start=${String(config.startLayer)}, end=${String(config.endLayer)})`,
      );
    }
    this.startLayer = config.startLayer;
    this.endLayer = config.endLayer;
    this.role = config.role;
    this.maxBatchSize = config.maxBatchSize ?? 4;
    this.phase = config.phase;
    this.url = config.endpoint.replace(/\/+$/, '');
    this.bearerToken = config.bearerToken ?? 'mesh-bench';
    this.connectTimeoutMs = config.connectTimeoutMs ?? 10_000;
    this.coalesceEnabled = readCoalesceFlag();
    // Fallback uses the same HTTPS endpoint for non-forward methods.
    this.httpsFallback = new Station({
      endpoint: config.endpoint,
      startLayer: config.startLayer,
      endLayer: config.endLayer,
      role: config.role,
      maxBatchSize: config.maxBatchSize,
    });
  }

  /**
   * Lazily open the WS socket. Idempotent: concurrent callers share the
   * same connect promise. Called from `forward()` on first use.
   */
  private async connect(): Promise<void> {
    if (this.socket && this.codec) return;
    if (this.connectPromise) return this.connectPromise;

    this.connectPromise = (async () => {
      const wsUrl = this.url.replace(/^https:/, 'wss:').replace(/^http:/, 'ws:') + '/flow';
      // Node.js WebSocket accepts headers in opts; browser WebSocket does
      // not. We try with headers first; if the runtime ignores them the
      // server will respond 401 and we surface the error to the caller.
      const ws: WebSocket = new (globalThis as unknown as { WebSocket: typeof WebSocket }).WebSocket(
        wsUrl,
        { headers: { Authorization: `Bearer ${this.bearerToken}` } },
      );
      (ws as unknown as { binaryType: string }).binaryType = 'arraybuffer';
      const codec = await FlowCodec.create({ wasmMode: 'off' });

      await new Promise<void>((resolve, reject) => {
        const timer = setTimeout(
          () => reject(new Error(`WSStation: connect timeout (${this.connectTimeoutMs}ms) ${wsUrl}`)),
          this.connectTimeoutMs,
        );
        ws.addEventListener('open', () => { clearTimeout(timer); resolve(); }, { once: true });
        ws.addEventListener('error', (e: Event) => {
          clearTimeout(timer);
          reject(new Error(`WSStation: connect error ${wsUrl} (${String(e)})`));
        }, { once: true });
      });

      ws.addEventListener('message', (evt: MessageEvent) => {
        try {
          const data = evt.data instanceof ArrayBuffer
            ? new Uint8Array(evt.data)
            : evt.data instanceof Uint8Array
              ? evt.data
              : null;
          if (!data) return;
          const tick = codec.decode(data).frame;
          const slot = this.pending.get(tick.sequence);
          if (!slot) return;
          this.pending.delete(tick.sequence);
          if ((tick.flags & 0x08 /* VENT */) !== 0) {
            const msg = new TextDecoder().decode(tick.payload);
            slot.reject(new Error(`WSStation /flow VENT: ${msg}`));
          } else if ((tick.flags & 0x10 /* FIN */) !== 0) {
            slot.resolve(bytesToFloat32(tick.payload));
          }
        } catch (err) {
          // Decode failure shouldn't kill the socket; reject the oldest
          // pending caller so the dispatcher gets a real error.
          const firstSeq = this.pending.keys().next().value as number | undefined;
          if (firstSeq !== undefined) {
            const slot = this.pending.get(firstSeq);
            this.pending.delete(firstSeq);
            slot?.reject(new Error(`WSStation: frame decode threw: ${(err as Error).message}`));
          }
        }
      });
      ws.addEventListener('close', () => this.failAllPending('socket closed'));
      ws.addEventListener('error', () => this.failAllPending('socket error'));

      this.socket = ws;
      this.codec = codec;
      // Leg 1 of the Mesh Consumption Wave (TOPOLOGY_CEILING_BREAKERS §4):
      // coalesce same-microtask `forward()` sends into one socket.send()
      // when GNOSIS_FRAME_COALESCE is on (default). The batcher validates
      // each payload as a single flow frame and concatenates queued frames
      // at end-of-microtask via the FlowSendBatcher's existing flush.
      if (this.coalesceEnabled) {
        this.sendBatcher = new FlowSendBatcher((payload: Uint8Array) => {
          // Copy into an ArrayBuffer-backed view: `payload.buffer` may be a
          // SharedArrayBuffer; lib.dom `WebSocket#send` rejects that shape.
          const copy = new Uint8Array(payload.byteLength);
          copy.set(payload);
          ws.send(copy.buffer);
        });
      } else {
        this.sendBatcher = null;
      }
    })();

    try {
      await this.connectPromise;
    } catch (err) {
      this.connectPromise = null;
      throw err;
    }
  }

  private failAllPending(reason: string): void {
    for (const [seq, slot] of this.pending) {
      slot.reject(new Error(`WSStation: ${reason} (seq=${seq})`));
    }
    this.pending.clear();
    this.socket = null;
    this.codec = null;
    this.connectPromise = null;
    if (this.sendBatcher) {
      // The underlying socket is gone; closing the batcher prevents a
      // pending flush from invoking sendNow on the dead socket.
      try { this.sendBatcher.close(); } catch { /* ignore */ }
      this.sendBatcher = null;
    }
  }

  /**
   * Send one PHASE_{A,B,C} frame on the persistent socket and await the
   * matching FIN response. Mirrors `bench-trisplit.ts` FlowClient.call().
   *
   * Note: the IStation contract is `residual → { residual }`. The WS
   * handler returns whatever the kernel produces for the configured
   * phase (split-a returns 2× hidden, split-b returns intermediate,
   * split-c returns 1× hidden). The Pipeline does not validate sizes
   * (only divisibility-by-4), so this is consistent with the HTTPS path
   * the bench script measured.
   */
  async forward(args: {
    residual: Float32Array;
    startPos: number;
    sequenceLength: number;
    requestId?: string;
  }): Promise<{
    residual: Float32Array;
    teleportation?: { layerDeficits: number[]; rounds: number };
  }> {
    if (!Number.isInteger(args.startPos) || args.startPos < 0) {
      throw new Error(`WSStation.forward: startPos must be >= 0 (got ${String(args.startPos)})`);
    }
    if (!Number.isInteger(args.sequenceLength) || args.sequenceLength <= 0) {
      throw new Error(
        `WSStation.forward: sequenceLength must be a positive integer (got ${String(args.sequenceLength)})`,
      );
    }
    await this.connect();
    if (!this.socket || !this.codec) {
      throw new Error('WSStation.forward: socket unavailable after connect()');
    }

    const phaseFlag = PHASE_FOR[this.phase] & PHASE_MASK;
    const streamId = STREAM_BASE_FOR[this.phase] + this.startLayer;
    const sequence = (this.nextSequence++ >>> 0);
    const bodyBytes = new Uint8Array(
      args.residual.buffer,
      args.residual.byteOffset,
      args.residual.byteLength,
    );
    const payload = encodeResidualPayload(args.startPos, args.sequenceLength, bodyBytes);
    const frame = this.codec.encode({
      streamId,
      sequence,
      flags: phaseFlag,
      payload,
    });

    const t0 = Date.now();
    const result = await new Promise<Float32Array>((resolve, reject) => {
      this.pending.set(sequence, { resolve, reject });
      try {
        if (this.sendBatcher) {
          // Batcher takes a Uint8Array view of the single flow frame and
          // coalesces same-microtask sends into one socket.send(). It does
          // an internal copy so we can pass the codec's view directly.
          this.sendBatcher.send(frame);
        } else {
          const copy = new Uint8Array(frame.byteLength);
          copy.set(frame);
          this.socket!.send(copy.buffer);
        }
      } catch (err) {
        this.pending.delete(sequence);
        reject(err as Error);
      }
    });
    const ms = Date.now() - t0;

    trace(
      `[WSStation:forward] url=${this.url} phase=${this.phase} layer=${this.startLayer}` +
        ` seq=${sequence} bytes_in=${bodyBytes.byteLength} bytes_out=${result.byteLength} ms=${ms}`,
    );

    return { residual: result };
  }

  // ── Cold-path delegation to HTTPS Station ─────────────────────────────
  // The /flow worker route does NOT yet handle these. Calls fall through
  // to the inner HTTPS Station so dispatchers can use a single station
  // object regardless of method.

  async embed(args: { tokens: Uint32Array; startPos: number; requestId?: string }): Promise<Float32Array> {
    return this.httpsFallback.embed(args);
  }

  async lmHead(args: { residual: Float32Array; sequenceLength: number; requestId?: string }): Promise<Float32Array> {
    return this.httpsFallback.lmHead(args);
  }

  async sampleNext(args: { residual: Float32Array; sequenceLength: number; requestId?: string }): Promise<number> {
    return this.httpsFallback.sampleNext(args);
  }

  async decodeNext(args: { tokenId: number; position: number; requestId?: string }): Promise<number> {
    return this.httpsFallback.decodeNext(args);
  }

  async cobordismEmbed(args: {
    tokens: Uint32Array;
    startPos: number;
    vocabStart: number;
    vocabEnd: number;
    requestId?: string;
  }): Promise<Float32Array> {
    return this.httpsFallback.cobordismEmbed(args);
  }

  async cobordismLmHead(args: {
    residual: Float32Array;
    sequenceLength: number;
    vocabStart: number;
    vocabEnd: number;
    requestId?: string;
  }): Promise<Float32Array> {
    return this.httpsFallback.cobordismLmHead(args);
  }

  async bizarroForward(args: {
    residual: Float32Array;
    startPos: number;
    sequenceLength: number;
    targetPosition: number;
    requestId?: string;
  }): Promise<{ rejectMask: number[] } | null> {
    return this.httpsFallback.bizarroForward(args);
  }

  async saveKvSession(args: {
    sessionId: string;
    prefixLength: number;
    requestId?: string;
  }): Promise<boolean> {
    return this.httpsFallback.saveKvSession(args);
  }

  async loadKvSession(args: {
    sessionId: string;
    requestId?: string;
  }): Promise<{ prefixLength: number } | null> {
    return this.httpsFallback.loadKvSession(args);
  }

  async deleteKvSession(args: {
    sessionId: string;
    requestId?: string;
  }): Promise<boolean> {
    return this.httpsFallback.deleteKvSession(args);
  }

  /**
   * Amplituhedron volume-cache replay. POSTs to `/amplituhedron/replay` on
   * the same HTTPS endpoint the socket was opened against (not the /flow WS
   * route — the worker's amplituhedron admin endpoints are HTTP-only per
   * AMPLITUHEDRON_PLAN.md §4). Request key tuple is JSON; response is the
   * bitwise binary envelope decoded via `decodeReplayResponse`. Returns
   * `{ hit: false }` on 204 No Content (miss), the decoded envelope on
   * 200, and `{ hit: false }` defensively on any other status / decode
   * error so a flaky admin route can never poison the coordinator's hot
   * path.
   */
  async amplituhedronReplay(args: {
    prefixHash: string;
    prefixLen: number;
    layerLo: number;
    layerHi: number;
    requestId?: string;
  }): Promise<{
    hit: boolean;
    tailResidual?: Float32Array;
    tailFingerprint?: string;
    prefixResidual?: Float32Array;
    prefixFingerprint?: string;
  }> {
    const url = `${this.url}/amplituhedron/replay`;
    const body = JSON.stringify({
      prefix_hash: args.prefixHash,
      prefix_len: args.prefixLen,
      layer_lo: args.layerLo,
      layer_hi: args.layerHi,
      ...(args.requestId !== undefined ? { request_id: args.requestId } : {}),
    });
    let res: Response;
    try {
      res = await fetch(url, {
        method: 'POST',
        body,
        headers: {
          'content-type': 'application/json',
          accept: AMPLITUHEDRON_CONTENT_TYPE,
          authorization: `Bearer ${this.bearerToken}`,
        },
      });
    } catch {
      // Network errors fall back to MISS — the coordinator handles this by
      // running a normal prefill. Never throw on the replay path.
      return { hit: false };
    }
    if (res.status === 204) return { hit: false };
    if (res.status === 200) {
      try {
        const buf = await res.arrayBuffer();
        return decodeReplayResponse(new Uint8Array(buf));
      } catch {
        return { hit: false };
      }
    }
    return { hit: false };
  }

  /**
   * Fire-and-forget capture POST. The coordinator never awaits this (see
   * Pipeline.scheduleAmplituhedronCapture → pipeline.ts:964), so we swallow
   * every error here: a failed capture just means the next same-prefix call
   * misses and re-runs prefill — a perf regression, not a correctness bug.
   * Body uses the bitwise codec in ./amplituhedron-wire.
   */
  async amplituhedronCapture(args: {
    prefixHash: string;
    prefixLen: number;
    layerLo: number;
    layerHi: number;
    tailResidual?: Float32Array;
    requestId?: string;
  }): Promise<void> {
    // Contract gap: IStation declares tailResidual optional, codec requires
    // non-optional Float32Array. No residual = nothing to capture.
    if (!args.tailResidual) {
      return;
    }
    const url = `${this.url}/amplituhedron/capture`;
    let body: Uint8Array;
    try {
      body = encodeCaptureRequest({
        prefixHash: args.prefixHash,
        prefixLen: args.prefixLen,
        layerLo: args.layerLo,
        layerHi: args.layerHi,
        tailResidual: args.tailResidual,
      });
    } catch {
      return;
    }
    try {
      await fetch(url, {
        method: 'POST',
        body: body.buffer.slice(
          body.byteOffset,
          body.byteOffset + body.byteLength,
        ) as ArrayBuffer,
        headers: {
          'content-type': AMPLITUHEDRON_CONTENT_TYPE,
          authorization: `Bearer ${this.bearerToken}`,
        },
      });
    } catch {
      // Capture is best-effort. A failure here just means the next call
      // misses and re-runs prefill. Never throw.
    }
  }

  /** Close the persistent socket. Pending calls reject. */
  close(): void {
    if (this.socket) {
      try { this.socket.close(); } catch { /* already closed */ }
    }
    this.failAllPending('close() called');
  }
}

// ── helpers ────────────────────────────────────────────────────────────

function bytesToFloat32(buf: Uint8Array): Float32Array {
  if (buf.byteLength % 4 !== 0) {
    throw new Error(`WSStation: response byteLength ${buf.byteLength} not /4`);
  }
  // Aeon-flow header is 10 bytes; payload arrives offset 10 in the raw
  // frame. After codec.decode() we only see the payload here, but it may
  // still be offset 8 (inference prefix had startPos/seqLen) if the
  // server included the prefix in its response. The server sends the
  // bare payload (no inference prefix on responses), so we copy into a
  // fresh aligned buffer for safety regardless of view alignment.
  const copy = new ArrayBuffer(buf.byteLength);
  new Uint8Array(copy).set(buf);
  return new Float32Array(copy);
}
