// @a0n/.ignored_gnosis/distributed-inference-host/src/station-flow-udp.ts
/**
 * StationFlowUdp — UDP datagram cousin of StationFlow.
 *
 * Why this exists
 * ---------------
 * `station-flow.ts` shipped the canonical aeon-flow 10-byte binary
 * frame over a Unix Domain Socket — the lowest-latency local-host
 * transport. UDP is the cross-host (or shared-FD-impossible) cousin:
 * SAME wire format, only the framing wrap differs.
 *
 *   UDS = 4-byte length prefix + aeon-flow frame.
 *   UDP = ONE datagram == ONE aeon-flow frame. No length prefix.
 *
 * Diversity-by-design: pick UDS for same-host (lowest latency, shipped
 * 2.15× HTTP), pick UDP for cross-host or when a UDS path / shared-FD
 * isn't available (different machines, different network namespaces,
 * cloud peers, ...). See docs/pneuma/mesh-flow-udp.md.
 *
 * Reuses the wire codec verbatim
 * ------------------------------
 * encodeFlowFrame / decodeFlowFrame / encodePayload / decodePayload
 * are lifted from station-flow.ts byte-for-byte. The ONLY transport
 * difference is `sendFrameUdp` vs `sendFrame` — a `node:dgram` socket
 * with chunked-reassembly logic for oversized responses.
 *
 * Multi-datagram chunking (RX side)
 * ---------------------------------
 * If the server sets the FLAG_CHUNK bit (0x04) on a response frame,
 * the payload begins with `[chunk_idx u16 BE | chunk_total u16 BE]`
 * followed by a slice of the canonical (meta_len-prefixed) payload.
 * The client reassembles by concatenating slices in chunk_idx order,
 * keyed by `sequence` (= request id). FIN (0x10) is set on the last
 * chunk. Once all chunks arrive, the result is decoded as a normal
 * single-datagram payload.
 *
 * Out-of-order arrivals: UDP reorders, so we hold a per-sequence
 * reassembly buffer keyed by chunk_idx and emit when chunk_total
 * distinct chunks have landed. Lost chunks surface as a per-call
 * timeout (default 30s, same as StationFlow).
 *
 * Runtime: requires Node `node:dgram`. CF Workers have no UDP send,
 * so this transport is host-side only by construction.
 */

import { createSocket, type Socket as DgramSocket } from 'node:dgram';
import type { IStation } from './types';

// ── Aeon-flow constants (mirror open-source/aeon/src/flow/FlowCodec.ts) ──
export const FLOW_HEADER_SIZE = 10;
export const FLOW_MAX_PAYLOAD = 0xffffff;
export const FLAG_FIN = 0x10;
export const FLAG_VENT = 0x08;
export const FLAG_RESPONSE = 0x20;
/** Set by the server when a response is split across multiple datagrams.
 *  The payload of every chunk frame begins with `[idx u16 BE | total u16 BE]`. */
export const FLAG_CHUNK = 0x04;

// Route opcodes (mirror fat-station-flow-udp.rs ROUTE_* constants).
export const ROUTE_HEALTH = 0x0001;
export const ROUTE_EMBED = 0x0002;
export const ROUTE_FORWARD = 0x0003;
export const ROUTE_LM_HEAD = 0x0004;
export const ROUTE_COBORDISM_EMBED = 0x0005;
export const ROUTE_COBORDISM_LM_HEAD = 0x0006;

const DEFAULT_TIMEOUT_MS = 30_000;
/** Per-datagram payload cap. macOS enforces `net.inet.udp.maxdgram=9216`
 *  with EMSGSIZE on send (no IP fragmentation rescue), so we cap at 8 KiB
 *  and chunk anything larger using the same FLAG_CHUNK + [idx | total]
 *  prefix shape the server uses for its responses. The Pneuma worst-case
 *  request is seq_len=8 × hidden_dim=3584 × 4 ≈ 114 KiB which needs ~14
 *  chunks per call. */
const MAX_DATAGRAM_PAYLOAD = 8 * 1024;

export interface StationFlowUdpConfig {
  /** UDP host:port the server listens on. */
  host: string;
  port: number;
  /** Inclusive layer-range start. */
  startLayer: number;
  /** Exclusive layer-range end. */
  endLayer: number;
  /** Optional declared role for routing. */
  role?: 'entry' | 'mid' | 'exit' | 'cobordism';
  /** Per-call timeout in ms. Default 30s. */
  timeoutMs?: number;
}

interface FlowFrameDecoded {
  streamId: number;
  sequence: number;
  flags: number;
  payload: Uint8Array;
}

interface ParsedReply {
  streamId: number;
  sequence: number;
  flags: number;
  meta: Record<string, unknown>;
  body: Uint8Array;
}

export class StationFlowUdp implements IStation {
  readonly startLayer: number;
  readonly endLayer: number;
  readonly role: 'entry' | 'mid' | 'exit' | 'cobordism' | undefined;
  readonly maxBatchSize = 8;
  private readonly host: string;
  private readonly port: number;
  private readonly timeoutMs: number;
  private nextSequence = 1;

  constructor(config: StationFlowUdpConfig) {
    if (!Number.isInteger(config.startLayer) || config.startLayer < 0) {
      throw new Error(
        `StationFlowUdp: startLayer must be a non-negative integer (got ${String(config.startLayer)})`,
      );
    }
    if (!Number.isInteger(config.endLayer) || config.endLayer < config.startLayer) {
      throw new Error(
        `StationFlowUdp: endLayer must be an integer >= startLayer (got start=${String(config.startLayer)}, end=${String(config.endLayer)})`,
      );
    }
    if (!Number.isInteger(config.port) || config.port <= 0 || config.port > 65535) {
      throw new Error(
        `StationFlowUdp: port must be an integer in (0, 65535] (got ${String(config.port)})`,
      );
    }
    this.startLayer = config.startLayer;
    this.endLayer = config.endLayer;
    this.role = config.role;
    this.host = config.host;
    this.port = config.port;
    this.timeoutMs = config.timeoutMs ?? DEFAULT_TIMEOUT_MS;
  }

  /** Health probe — returns the server's parsed JSON status object. */
  async health(): Promise<unknown> {
    const reply = await this.call(ROUTE_HEALTH, {}, new Uint8Array(0));
    expectOk(reply, 'health');
    return reply.meta;
  }

  async forward(args: {
    residual: Float32Array;
    startPos: number;
    sequenceLength: number;
    requestId?: string;
  }): Promise<{
    residual: Float32Array;
    teleportation?: { layerDeficits: number[]; rounds: number };
  }> {
    if (!Number.isInteger(args.startPos) || args.startPos < 0) {
      throw new Error(
        `StationFlowUdp.forward: startPos must be >= 0 (got ${String(args.startPos)})`,
      );
    }
    if (!Number.isInteger(args.sequenceLength) || args.sequenceLength <= 0) {
      throw new Error(
        `StationFlowUdp.forward: sequenceLength must be a positive integer (got ${String(args.sequenceLength)})`,
      );
    }
    const reply = await this.call(
      ROUTE_FORWARD,
      {
        start_pos: args.startPos,
        seq_len: args.sequenceLength,
        layer_range: `${this.startLayer}-${this.endLayer}`,
        request_id: args.requestId ?? `req-${Date.now()}`,
      },
      float32ToBytes(args.residual),
    );
    expectOk(reply, 'forward');
    const residual = bytesToFloat32(reply.body, 'forward');
    const teleportation = parseTeleportationMeta(reply.meta);
    return teleportation ? { residual, teleportation } : { residual };
  }

  async embed(args: {
    tokens: Uint32Array;
    startPos: number;
    requestId?: string;
  }): Promise<Float32Array> {
    const reply = await this.call(
      ROUTE_EMBED,
      {
        start_pos: args.startPos,
        token_count: args.tokens.length,
        request_id: args.requestId ?? `req-${Date.now()}`,
      },
      uint32ToBytes(args.tokens),
    );
    expectOk(reply, 'embed');
    return bytesToFloat32(reply.body, 'embed');
  }

  async lmHead(args: {
    residual: Float32Array;
    sequenceLength: number;
    requestId?: string;
  }): Promise<Float32Array> {
    if (!Number.isInteger(args.sequenceLength) || args.sequenceLength <= 0) {
      throw new Error(
        `StationFlowUdp.lmHead: sequenceLength must be a positive integer (got ${String(args.sequenceLength)})`,
      );
    }
    const reply = await this.call(
      ROUTE_LM_HEAD,
      {
        seq_len: args.sequenceLength,
        request_id: args.requestId ?? `req-${Date.now()}`,
      },
      float32ToBytes(args.residual),
    );
    expectOk(reply, 'lm_head');
    return bytesToFloat32(reply.body, 'lm_head');
  }

  async cobordismEmbed(args: {
    tokens: Uint32Array;
    startPos: number;
    vocabStart: number;
    vocabEnd: number;
    requestId?: string;
  }): Promise<Float32Array> {
    const reply = await this.call(
      ROUTE_COBORDISM_EMBED,
      {
        start_pos: args.startPos,
        token_count: args.tokens.length,
        vocab_start: args.vocabStart,
        vocab_end: args.vocabEnd,
        request_id: args.requestId ?? `req-${Date.now()}`,
      },
      uint32ToBytes(args.tokens),
    );
    expectOk(reply, 'cobordism_embed');
    return bytesToFloat32(reply.body, 'cobordism_embed');
  }

  async cobordismLmHead(args: {
    residual: Float32Array;
    sequenceLength: number;
    vocabStart: number;
    vocabEnd: number;
    requestId?: string;
  }): Promise<Float32Array> {
    const reply = await this.call(
      ROUTE_COBORDISM_LM_HEAD,
      {
        seq_len: args.sequenceLength,
        vocab_start: args.vocabStart,
        vocab_end: args.vocabEnd,
        request_id: args.requestId ?? `req-${Date.now()}`,
      },
      float32ToBytes(args.residual),
    );
    expectOk(reply, 'cobordism_lm_head');
    return bytesToFloat32(reply.body, 'cobordism_lm_head');
  }

  async bizarroForward(_args: {
    residual: Float32Array;
    startPos: number;
    sequenceLength: number;
    targetPosition: number;
    requestId?: string;
  }): Promise<{ rejectMask: number[] } | null> {
    return null;
  }

  // ── Amplituhedron admin path: intentionally NOT implemented ────────
  //
  // Same rationale as StationFlow (see station-flow.ts): the JSON-shaped
  // /amplituhedron/{replay,capture} contract has no ROUTE_* opcode in
  // fat-station-flow-udp.rs. UDP additionally complicates capture's
  // tail_residual payload, which can easily blow past the 1500B MTU
  // (and the multi-datagram split path is only wired for request
  // bodies, not for arbitrary-sized JSON envelopes). Until both sides
  // gain a flow-native amplituhedron route, we omit these methods.
  // IStation marks them `?optional` and Pipeline guards with
  // `if (!capable.amplituhedronReplay)` before dispatch, so absence here
  // degrades cleanly to MISS + normal prefill.

  // ── transport ──────────────────────────────────────────────────────

  private async call(
    streamId: number,
    meta: Record<string, unknown>,
    body: Uint8Array,
  ): Promise<ParsedReply> {
    const sequence = (this.nextSequence++ >>> 0);
    const payload = encodePayload(meta, body);
    // If the canonical payload fits in one datagram, send a single
    // FLAG_CHUNK-clear frame. Otherwise split the payload into
    // (MAX_DATAGRAM_PAYLOAD - 4)-sized slices and send each as a
    // FLAG_CHUNK-set frame with a [idx u16 BE | total u16 BE] prefix.
    const datagrams = encodeRequestDatagrams(streamId, sequence, payload);
    const replyFrame = await sendFramesUdp(
      this.host,
      this.port,
      datagrams,
      sequence,
      this.timeoutMs,
    );
    if ((replyFrame.flags & FLAG_RESPONSE) === 0) {
      throw new Error(
        `StationFlowUdp: reply missing IS_RESPONSE flag (flags=0x${replyFrame.flags.toString(16)})`,
      );
    }
    if (replyFrame.streamId !== streamId) {
      throw new Error(
        `StationFlowUdp: reply streamId 0x${replyFrame.streamId.toString(16)} != request 0x${streamId.toString(16)}`,
      );
    }
    if (replyFrame.sequence !== sequence) {
      throw new Error(
        `StationFlowUdp: reply sequence ${replyFrame.sequence} != request ${sequence}`,
      );
    }
    const { meta: parsedMeta, body: parsedBody } = decodePayload(replyFrame.payload);
    return {
      streamId: replyFrame.streamId,
      sequence: replyFrame.sequence,
      flags: replyFrame.flags,
      meta: parsedMeta,
      body: parsedBody,
    };
  }
}

// ── aeon-flow codec (verbatim from station-flow.ts) ─────────────────

/**
 * Handles the gnosis encode Flow Frame workflow.
 */
export function encodeFlowFrame(frame: FlowFrameDecoded): Uint8Array {
  if (frame.payload.byteLength > FLOW_MAX_PAYLOAD) {
    throw new Error(
      `StationFlowUdp: payload ${frame.payload.byteLength}B exceeds u24 max ${FLOW_MAX_PAYLOAD}B`,
    );
  }
  const out = new Uint8Array(FLOW_HEADER_SIZE + frame.payload.byteLength);
  out[0] = (frame.streamId >>> 8) & 0xff;
  out[1] = frame.streamId & 0xff;
  out[2] = (frame.sequence >>> 24) & 0xff;
  out[3] = (frame.sequence >>> 16) & 0xff;
  out[4] = (frame.sequence >>> 8) & 0xff;
  out[5] = frame.sequence & 0xff;
  out[6] = frame.flags & 0xff;
  const len = frame.payload.byteLength;
  out[7] = (len >> 16) & 0xff;
  out[8] = (len >> 8) & 0xff;
  out[9] = len & 0xff;
  out.set(frame.payload, FLOW_HEADER_SIZE);
  return out;
}

/**
 * Handles the gnosis decode Flow Frame workflow.
 */
export function decodeFlowFrame(bytes: Uint8Array): FlowFrameDecoded {
  if (bytes.byteLength < FLOW_HEADER_SIZE) {
    throw new Error(
      `StationFlowUdp: frame ${bytes.byteLength}B < header ${FLOW_HEADER_SIZE}B`,
    );
  }
  const streamId = (bytes[0]! << 8) | bytes[1]!;
  const sequence =
    ((bytes[2]! * 0x1000000) + ((bytes[3]! << 16) | (bytes[4]! << 8) | bytes[5]!)) >>> 0;
  const flags = bytes[6]!;
  const length = (bytes[7]! << 16) | (bytes[8]! << 8) | bytes[9]!;
  if (bytes.byteLength < FLOW_HEADER_SIZE + length) {
    throw new Error(
      `StationFlowUdp: truncated payload — declared=${length}B, available=${bytes.byteLength - FLOW_HEADER_SIZE}B`,
    );
  }
  return {
    streamId,
    sequence,
    flags,
    payload: bytes.subarray(FLOW_HEADER_SIZE, FLOW_HEADER_SIZE + length),
  };
}

function encodePayload(meta: Record<string, unknown>, body: Uint8Array): Uint8Array {
  const metaJson = JSON.stringify(meta);
  const metaBytes = new TextEncoder().encode(metaJson);
  if (metaBytes.byteLength > 0xffff) {
    throw new Error(
      `StationFlowUdp: meta_json ${metaBytes.byteLength}B exceeds u16 max 65535`,
    );
  }
  const out = new Uint8Array(2 + metaBytes.byteLength + body.byteLength);
  out[0] = (metaBytes.byteLength >> 8) & 0xff;
  out[1] = metaBytes.byteLength & 0xff;
  out.set(metaBytes, 2);
  out.set(body, 2 + metaBytes.byteLength);
  return out;
}

function decodePayload(payload: Uint8Array): {
  meta: Record<string, unknown>;
  body: Uint8Array;
} {
  if (payload.byteLength < 2) {
    throw new Error('StationFlowUdp: payload < 2B (no meta-length prefix)');
  }
  const metaLen = (payload[0]! << 8) | payload[1]!;
  if (payload.byteLength < 2 + metaLen) {
    throw new Error(
      `StationFlowUdp: payload ${payload.byteLength}B < 2 + meta_len ${metaLen}`,
    );
  }
  const metaBytes = payload.subarray(2, 2 + metaLen);
  const metaJson = new TextDecoder().decode(metaBytes);
  let meta: Record<string, unknown>;
  try {
    meta = JSON.parse(metaJson) as Record<string, unknown>;
  } catch (e) {
    throw new Error(`StationFlowUdp: meta json parse: ${(e as Error).message}`);
  }
  const body = payload.subarray(2 + metaLen);
  return { meta, body };
}

function expectOk(reply: ParsedReply, label: string): void {
  if ((reply.flags & FLAG_VENT) !== 0) {
    const errMsg =
      typeof reply.meta['error'] === 'string' ? (reply.meta['error'] as string) : 'unknown error';
    const status =
      typeof reply.meta['status'] === 'number' ? (reply.meta['status'] as number) : 500;
    throw new Error(`StationFlowUdp ${label}: status ${status} — ${errMsg}`);
  }
  const status = reply.meta['status'];
  if (typeof status === 'number' && status !== 200) {
    throw new Error(`StationFlowUdp ${label}: status ${status} — ${JSON.stringify(reply.meta)}`);
  }
}

function float32ToBytes(arr: Float32Array): Uint8Array {
  return new Uint8Array(arr.buffer, arr.byteOffset, arr.byteLength).slice();
}

function uint32ToBytes(arr: Uint32Array): Uint8Array {
  return new Uint8Array(arr.buffer, arr.byteOffset, arr.byteLength).slice();
}

function bytesToFloat32(buf: Uint8Array, label: string): Float32Array {
  if (buf.byteLength % 4 !== 0) {
    throw new Error(
      `StationFlowUdp ${label}: response byteLength ${buf.byteLength} is not a multiple of 4`,
    );
  }
  const copy = new ArrayBuffer(buf.byteLength);
  new Uint8Array(copy).set(buf);
  return new Float32Array(copy);
}

function parseTeleportationMeta(
  meta: Record<string, unknown>,
): { layerDeficits: number[]; rounds: number } | undefined {
  const def = meta['teleport_deficits'];
  const rnd = meta['teleport_rounds'];
  if (!Array.isArray(def) || typeof rnd !== 'number') return undefined;
  const layerDeficits = def
    .map((v) => (typeof v === 'number' ? v : Number.parseInt(String(v), 10)))
    .filter((v) => Number.isFinite(v));
  if (layerDeficits.length !== def.length) return undefined;
  return { layerDeficits, rounds: rnd };
}

// ── Node dgram send/recv with chunk reassembly ──────────────────────

interface ReassemblyBucket {
  total: number;
  /** chunk_idx -> slice. Sparse until all chunks land. */
  chunks: Map<number, Uint8Array>;
  /** Cached header from the FIRST chunk we saw — every chunk shares
   *  stream_id, sequence and (modulo CHUNK/FIN bits) flags. */
  streamId: number;
  flags: number;
}

/** Encode a request payload into one or more outbound datagrams.
 *  Single-datagram path: FLAG_CHUNK clear, payload as-is.
 *  Multi-datagram path: FLAG_CHUNK set on every frame, payload prefixed
 *  with `[idx u16 BE | total u16 BE]`. */
function encodeRequestDatagrams(
  streamId: number,
  sequence: number,
  payload: Uint8Array,
): Uint8Array[] {
  if (payload.byteLength <= MAX_DATAGRAM_PAYLOAD) {
    return [
      encodeFlowFrame({ streamId, sequence, flags: 0, payload }),
    ];
  }
  const sliceSize = MAX_DATAGRAM_PAYLOAD - 4;
  const total = Math.ceil(payload.byteLength / sliceSize);
  if (total > 0xffff) {
    throw new Error(
      `StationFlowUdp: request payload requires ${total} chunks (> u16 max)`,
    );
  }
  const out: Uint8Array[] = new Array(total);
  for (let i = 0; i < total; i++) {
    const start = i * sliceSize;
    const end = Math.min(start + sliceSize, payload.byteLength);
    const slice = payload.subarray(start, end);
    const chunkPayload = new Uint8Array(4 + slice.byteLength);
    chunkPayload[0] = (i >> 8) & 0xff;
    chunkPayload[1] = i & 0xff;
    chunkPayload[2] = (total >> 8) & 0xff;
    chunkPayload[3] = total & 0xff;
    chunkPayload.set(slice, 4);
    out[i] = encodeFlowFrame({
      streamId,
      sequence,
      flags: FLAG_CHUNK,
      payload: chunkPayload,
    });
  }
  return out;
}

function sendFramesUdp(
  host: string,
  port: number,
  frames: Uint8Array[],
  expectedSequence: number,
  timeoutMs: number,
): Promise<FlowFrameDecoded> {
  return new Promise((resolve, reject) => {
    const socket: DgramSocket = createSocket('udp4');
    let settled = false;
    const bucket: ReassemblyBucket = {
      total: -1,
      chunks: new Map(),
      streamId: 0,
      flags: 0,
    };

    const finish = (err: Error | null, value?: FlowFrameDecoded): void => {
      if (settled) return;
      settled = true;
      if (timer) clearTimeout(timer);
      try {
        socket.close();
      } catch {
        // ignore
      }
      if (err) reject(err);
      else resolve(value!);
    };

    const timer = setTimeout(() => {
      finish(new Error(`StationFlowUdp: call timed out after ${timeoutMs}ms`));
    }, timeoutMs);

    socket.on('error', (err: Error) => finish(err));

    socket.on('message', (msg: Buffer) => {
      let decoded: FlowFrameDecoded;
      try {
        decoded = decodeFlowFrame(new Uint8Array(msg));
      } catch (e) {
        finish(new Error(`StationFlowUdp: decode datagram: ${(e as Error).message}`));
        return;
      }
      // Demux: ignore datagrams for other in-flight requests (we use one
      // socket per call so this is mostly defensive).
      if (decoded.sequence !== expectedSequence) {
        return;
      }
      if ((decoded.flags & FLAG_CHUNK) === 0) {
        // Single-datagram response — done.
        finish(null, decoded);
        return;
      }
      // Chunked response: peel off [idx u16 BE | total u16 BE] prefix
      // and stash the slice. Emit when all `total` chunks have landed.
      const p = decoded.payload;
      if (p.byteLength < 4) {
        finish(new Error(`StationFlowUdp: chunk payload < 4B header (${p.byteLength}B)`));
        return;
      }
      const idx = (p[0]! << 8) | p[1]!;
      const total = (p[2]! << 8) | p[3]!;
      if (total === 0) {
        finish(new Error(`StationFlowUdp: chunk total=0 (invalid)`));
        return;
      }
      if (idx >= total) {
        finish(new Error(`StationFlowUdp: chunk idx ${idx} >= total ${total}`));
        return;
      }
      if (bucket.total < 0) {
        bucket.total = total;
        bucket.streamId = decoded.streamId;
        bucket.flags = decoded.flags;
      } else if (bucket.total !== total) {
        finish(
          new Error(
            `StationFlowUdp: chunk total mismatch — bucket=${bucket.total}, frame=${total}`,
          ),
        );
        return;
      }
      // Last chunk carries FIN — accumulate the FIN bit into the
      // reassembled frame's flags.
      if ((decoded.flags & FLAG_FIN) !== 0) bucket.flags |= FLAG_FIN;
      // Strip the CHUNK bit from the reassembled frame's flags — the
      // reconstructed frame is no longer chunked.
      bucket.chunks.set(idx, p.subarray(4));
      if (bucket.chunks.size < bucket.total) return;

      // Concatenate in idx order.
      let totalLen = 0;
      for (let i = 0; i < bucket.total; i++) {
        const s = bucket.chunks.get(i);
        if (!s) {
          finish(new Error(`StationFlowUdp: missing chunk idx=${i} of ${bucket.total}`));
          return;
        }
        totalLen += s.byteLength;
      }
      const reassembled = new Uint8Array(totalLen);
      let off = 0;
      for (let i = 0; i < bucket.total; i++) {
        const s = bucket.chunks.get(i)!;
        reassembled.set(s, off);
        off += s.byteLength;
      }
      finish(null, {
        streamId: bucket.streamId,
        sequence: expectedSequence,
        flags: bucket.flags & ~FLAG_CHUNK,
        payload: reassembled,
      });
    });

    // Bind to ephemeral port, then send each datagram in order. Bind
    // explicitly so the 'message' handler is wired before the server
    // can reply.
    const sendNext = (idx: number): void => {
      if (idx >= frames.length) return;
      const f = frames[idx]!;
      socket.send(f, 0, f.byteLength, port, host, (err) => {
        if (err) {
          finish(err);
          return;
        }
        // Send the next frame on the same tick. UDP is lossy and
        // unordered by spec but on localhost the kernel preserves
        // order; for cross-host the server's reassembler keys on
        // chunk_idx so order doesn't matter.
        sendNext(idx + 1);
      });
    };
    socket.bind(0, () => {
      sendNext(0);
    });
  });
}
