// @a0n/.ignored_gnosis/distributed-inference-host/src/station.ts
/**
 * Station — TS-side adapter that fronts a single CF Worker pinned to a
 * contiguous transformer layer range.
 *
 * Wire shape mirrors `open-source/aether/src/gcp-layer-node.ts`:
 *   - request body  : raw Float32Array bytes (octet-stream), zerocopy
 *   - response body : raw Float32Array bytes (octet-stream)
 *   - metadata      : compact ASCII headers (X-Position, X-Sequence-Length,
 *                     X-Layer-Range, X-Request-Id)
 *
 * Endpoints:
 *   POST /forward   — run forward_range(residual, startLayer, endLayer,
 *                     startPos, sequenceLength) on the remote Worker
 *   POST /embed     — entry-only: tokens (Uint32Array) -> initial residual
 *                     (Float32Array)
 *   POST /lm-head   — exit-only: residual -> logits
 *
 * Station is a thin RPC client. Pipelining and chunk dispatch live in
 * `./pipeline.ts`. The Station never holds in-flight Promise state of its own.
 */

import {
  AMPLITUHEDRON_CONTENT_TYPE,
  encodeCaptureRequest,
  decodeReplayResponse,
} from './amplituhedron-wire';
import { requireBuleHotReplicaAdmission } from './cost-algebra-enforcement';
import type {
  AckermannAdmissionGate,
  IStation,
  PersonalityStylePreferences,
  PersonalitySteeringSignature,
  PersonalitySteeringSpec,
} from './types';

export type {
  PersonalityStylePreferences,
  PersonalitySteeringSignature,
  PersonalitySteeringSpec,
};

export interface StationEndpoint {
  fetch(req: Request): Promise<Response>;
}

export interface StationConfig {
  /** URL or service-binding handle that resolves to a remote CF Worker. */
  endpoint: StationEndpoint | string;
  /** Inclusive layer-range start. */
  startLayer: number;
  /** Exclusive layer-range end. */
  endLayer: number;
  /** Optional: declared role for routing, "entry" | "mid" | "exit" | "cobordism". */
  role?: 'entry' | 'mid' | 'exit' | 'cobordism';
  /**
   * Optional explicit base URL used when `endpoint` is a service-binding-style
   * object (no URL of its own). Defaults to "http://station".
   */
  baseUrl?: string;
  /**
   * Optional override for the per-station chunk cap surfaced to the
   * orchestrator (`Pipeline` reads `IStation.maxBatchSize`). Defaults to the
   * wasm-safe value of 4 because `Station` may front a CF Worker built for
   * `wasm32` (where `BATCH_CAPACITY = 4`). Callers that KNOW their endpoint
   * is a native fat-station HTTP server may pass `maxBatchSize: 8` to recover
   * the ~2× prefill round-trip win. See docs/pneuma/per-station-batch-cap.md.
   */
  maxBatchSize?: number;
}

function applyAckermannAdmissionGate(
  headers: Record<string, string>,
  gate: AckermannAdmissionGate | undefined
): void {
  if (!gate) return;
  const decision = requireBuleHotReplicaAdmission(
    gate.projection,
    gate.replicas,
    gate.ceiling
  );
  headers['X-Ackermann-Admission'] = decision.admitted
    ? 'admitted'
    : 'rejected';
  headers['X-Ackermann-Heat'] = String(decision.totalHeat);
  headers['X-Ackermann-Ceiling'] = String(decision.ceiling);
}

function formatFetchError(error: unknown): string {
  if (!(error instanceof Error)) return String(error);
  const cause = (error as Error & { cause?: unknown }).cause;
  if (cause instanceof Error) {
    const code = (cause as Error & { code?: string }).code;
    return `${error.message}; cause=${cause.name}: ${cause.message}${code ? ` code=${code}` : ''}`;
  }
  return error.message;
}

export class Station implements IStation {
  readonly startLayer: number;
  readonly endLayer: number;
  readonly role: 'entry' | 'mid' | 'exit' | 'cobordism' | undefined;
  // Station fronts both native fat-station HTTP servers and wasm CF Workers;
  // defaults to wasm-safe 4. Pass `maxBatchSize: 8` via constructor for
  // known-native targets.
  readonly maxBatchSize: number;

  private readonly endpoint: StationEndpoint;
  private readonly baseUrl: string;

  constructor(config: StationConfig) {
    if (!Number.isInteger(config.startLayer) || config.startLayer < 0) {
      throw new Error(
        `Station: startLayer must be a non-negative integer (got ${String(
          config.startLayer
        )})`
      );
    }
    if (
      !Number.isInteger(config.endLayer) ||
      config.endLayer < config.startLayer
    ) {
      throw new Error(
        `Station: endLayer must be an integer >= startLayer (got start=${String(
          config.startLayer
        )}, end=${String(config.endLayer)})`
      );
    }
    this.startLayer = config.startLayer;
    this.endLayer = config.endLayer;
    this.role = config.role;
    if (config.maxBatchSize !== undefined) {
      if (!Number.isInteger(config.maxBatchSize) || config.maxBatchSize <= 0) {
        throw new Error(
          `Station: maxBatchSize must be a positive integer when set (got ${String(
            config.maxBatchSize
          )})`
        );
      }
      this.maxBatchSize = config.maxBatchSize;
    } else {
      this.maxBatchSize = 4;
    }

    if (typeof config.endpoint === 'string') {
      const base = config.endpoint;
      this.baseUrl = base.replace(/\/+$/, '');
      this.endpoint = { fetch: (req) => fetch(req) };
    } else {
      this.endpoint = config.endpoint;
      this.baseUrl = (config.baseUrl ?? 'http://station').replace(/\/+$/, '');
    }
  }

  /**
   * Run forward_range(residual, startLayer, endLayer, startPos, sequenceLength)
   * on the remote Worker. Returns the post-range residual as a Float32Array.
   *
   * `residual.length` MUST equal `hiddenDim * sequenceLength`. The Station
   * does not know `hiddenDim` and so cannot validate this -- the remote node
   * (or the test fetcher) is the authority.
   */
  async forward(args: {
    residual: Float32Array;
    startPos: number;
    sequenceLength: number;
    requestId?: string;
  }): Promise<{
    residual: Float32Array;
    teleportation?: { layerDeficits: number[]; rounds: number };
  }> {
    if (!Number.isInteger(args.startPos) || args.startPos < 0) {
      throw new Error(
        `Station.forward: startPos must be >= 0 (got ${String(args.startPos)})`
      );
    }
    if (!Number.isInteger(args.sequenceLength) || args.sequenceLength <= 0) {
      throw new Error(
        `Station.forward: sequenceLength must be a positive integer (got ${String(
          args.sequenceLength
        )})`
      );
    }
    const body = floatArrayToBody(args.residual);
    const headers: Record<string, string> = {
      'Content-Type': 'application/octet-stream',
      'X-Position': String(args.startPos),
      'X-Sequence-Length': String(args.sequenceLength),
      'X-Layer-Range': `${this.startLayer}-${this.endLayer}`,
      'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
    };
    let res: Response;
    try {
      res = await this.endpoint.fetch(
        new Request(`${this.baseUrl}/forward`, {
          method: 'POST',
          headers,
          body,
        })
      );
    } catch (error) {
      throw new Error(
        `Station.forward fetch failed url=${this.baseUrl}/forward layers=${this.startLayer}-${this.endLayer} startPos=${args.startPos} sequenceLength=${args.sequenceLength}: ${formatFetchError(error)}`,
        { cause: error },
      );
    }

    const residual = await readFloatArrayResponse(res, '/forward');
    const teleportation = parseTeleportationHeaders(res.headers);

    return { residual, teleportation };
  }

  /**
   * Entry-only: produce the embedding for a chunk of tokens and return the
   * initial residual stream as a Float32Array of shape `[N * hiddenDim]`.
   */
  async embed(args: {
    tokens: Uint32Array;
    startPos: number;
    requestId?: string;
  }): Promise<Float32Array> {
    if (this.role !== undefined && this.role !== 'entry') {
      throw new Error(
        `Station.embed: only entry stations may embed (this station role=${this.role})`
      );
    }
    const body = uintArrayToBody(args.tokens);
    const headers: Record<string, string> = {
      'Content-Type': 'application/octet-stream',
      'X-Position': String(args.startPos),
      'X-Token-Count': String(args.tokens.length),
      'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
    };
    let res: Response;
    try {
      res = await this.endpoint.fetch(
        new Request(`${this.baseUrl}/embed`, {
          method: 'POST',
          headers,
          body,
        })
      );
    } catch (error) {
      throw new Error(
        `Station.embed fetch failed url=${this.baseUrl}/embed startPos=${args.startPos} tokenCount=${args.tokens.length}: ${formatFetchError(error)}`,
        { cause: error },
      );
    }
    return await readFloatArrayResponse(res, '/embed');
  }

  /**
   * Fire-and-forget speculative reject-mask forward. Mirrors aether's bizarro
   * companion path (`open-source/aether/src/gcp-coordinator.ts` ::
   * `dispatchBizarroCompanion`): instead of computing the next token, this
   * call asks the layer node to compute a sparse top-k set of vocabulary
   * indices the model would clearly reject for the target position. The
   * coordinator stashes the mask and uses it for early retirement when
   * `argmax(nominal_logits) ∉ rejectMask`.
   *
   * Wire format:
   *   - Same `/forward` endpoint with `X-Forward-Mode: bizarro` (matches
   *     `gcp-layer-node.ts` line 3295's `modeHeader === 'bizarro'` branch).
   *   - Optional `X-Target-Position` for the position whose mask we want.
   *   - Reply carries the reject mask as a JSON-encoded `number[]` in the
   *     `X-Reject-Mask` response header (the Worker doesn't need to support
   *     this yet -- if absent, this method returns null and the caller treats
   *     the dispatch as a no-op). This piggybacks on aether's existing
   *     header-as-JSON pattern (`X-Layer-Times`, `X-KV-Cache-Stats`).
   *
   * Errors are swallowed silently. Speculative work must NEVER poison the
   * nominal pipeline -- a 4xx, 5xx, network failure, or malformed mask
   * response must all degrade to "no mask" without throwing to the caller.
   *
   * Returns null when no usable mask is available (error, missing header,
   * malformed JSON, or empty array).
   */
  async bizarroForward(args: {
    residual: Float32Array;
    startPos: number;
    sequenceLength: number;
    targetPosition: number;
    requestId?: string;
  }): Promise<{ rejectMask: number[] } | null> {
    if (!Number.isInteger(args.startPos) || args.startPos < 0) {
      return null;
    }
    if (!Number.isInteger(args.sequenceLength) || args.sequenceLength <= 0) {
      return null;
    }
    if (!Number.isInteger(args.targetPosition) || args.targetPosition < 0) {
      return null;
    }
    try {
      const body = floatArrayToBody(args.residual);
      const headers: Record<string, string> = {
        'Content-Type': 'application/octet-stream',
        'X-Position': String(args.startPos),
        'X-Sequence-Length': String(args.sequenceLength),
        'X-Layer-Range': `${this.startLayer}-${this.endLayer}`,
        'X-Forward-Mode': 'bizarro',
        'X-Target-Position': String(args.targetPosition),
        'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
      };
      const res = await this.endpoint.fetch(
        new Request(`${this.baseUrl}/forward`, {
          method: 'POST',
          headers,
          body,
        })
      );
      if (!res.ok) return null;
      const maskHeader = res.headers.get('X-Reject-Mask');
      // Drain body so the connection can close cleanly even when we ignore it.
      try {
        await res.arrayBuffer();
      } catch {
        // ignore
      }
      if (!maskHeader) return null;
      const parsed = JSON.parse(maskHeader) as unknown;
      if (!Array.isArray(parsed)) return null;
      const rejectMask: number[] = [];
      for (const v of parsed) {
        if (
          typeof v === 'number' &&
          Number.isFinite(v) &&
          Number.isInteger(v) &&
          v >= 0
        ) {
          rejectMask.push(v);
        }
      }
      if (!rejectMask.length) return null;
      return { rejectMask };
    } catch {
      // Speculative — drop silently.
      return null;
    }
  }

  /** Exit-only: apply final-norm + lm_head, return logits for the last token. */
  async lmHead(args: {
    residual: Float32Array;
    sequenceLength: number;
    requestId?: string;
    ackermannAdmission?: AckermannAdmissionGate;
  }): Promise<Float32Array> {
    if (this.role !== undefined && this.role !== 'exit') {
      throw new Error(
        `Station.lmHead: only exit stations may run lm_head (this station role=${this.role})`
      );
    }
    if (!Number.isInteger(args.sequenceLength) || args.sequenceLength <= 0) {
      throw new Error(
        `Station.lmHead: sequenceLength must be a positive integer (got ${String(
          args.sequenceLength
        )})`
      );
    }
    const body = floatArrayToBody(args.residual);
    const headers: Record<string, string> = {
      'Content-Type': 'application/octet-stream',
      'X-Sequence-Length': String(args.sequenceLength),
      'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
    };
    applyAckermannAdmissionGate(headers, args.ackermannAdmission);
    const res = await this.endpoint.fetch(
      new Request(`${this.baseUrl}/lm-head`, {
        method: 'POST',
        headers,
        body,
      })
    );
    return await readFloatArrayResponse(res, '/lm-head');
  }

  /**
   * Exit-only: install a personality logit-steering filter on the station's
   * shared pipeline (the lm-head/logits live here). The station compiles a
   * per-token signed bias vector against its own vocab and applies it before the
   * argmax in `/decode-next`, gated by terminal prosody. Call once at the start
   * of a halogram session; `clearPersonalitySteering` restores the base model.
   */
  async installPersonalitySteering(
    spec: PersonalitySteeringSpec,
    requestId?: string
  ): Promise<void> {
    if (this.role !== undefined && this.role !== 'exit') {
      throw new Error(
        `Station.installPersonalitySteering: only exit stations carry the lm-head/logits (this station role=${this.role})`
      );
    }
    // Path A: a precompiled per-token vector (used when the station can't compile
    // from style_preferences against its own tokenizer). Otherwise Path B:
    // server-side compile from style_preferences + signature.
    const payload: Record<string, unknown> = spec.boundaryBias
      ? { boundary_bias: spec.boundaryBias }
      : { style_preferences: spec.stylePreferences, signature: spec.signature };
    if (spec.restScale !== undefined) payload.rest_scale = spec.restScale;
    const res = await this.endpoint.fetch(
      new Request(`${this.baseUrl}/personality-steering`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'X-Request-Id': requestId ?? `req-${Date.now()}`,
        },
        body: JSON.stringify(payload),
      })
    );
    if (!res.ok) {
      throw new Error(
        `Station.installPersonalitySteering: ${res.status} ${await res
          .text()
          .catch(() => '')}`
      );
    }
  }

  /** Exit-only: clear any installed personality steering (restore base-model argmax). */
  async clearPersonalitySteering(requestId?: string): Promise<void> {
    if (this.role !== undefined && this.role !== 'exit') return;
    const res = await this.endpoint.fetch(
      new Request(`${this.baseUrl}/personality-steering`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'X-Request-Id': requestId ?? `req-${Date.now()}`,
        },
        body: JSON.stringify({ clear: true }),
      })
    );
    if (!res.ok) {
      throw new Error(`Station.clearPersonalitySteering: ${res.status}`);
    }
  }

  /** Exit-only: apply final-norm + lm_head and return the greedy token id. */
  async sampleNext(args: {
    residual: Float32Array;
    sequenceLength: number;
    requestId?: string;
    ackermannAdmission?: AckermannAdmissionGate;
  }): Promise<number> {
    if (this.role !== undefined && this.role !== 'exit') {
      throw new Error(
        `Station.sampleNext: only exit stations may sample (this station role=${this.role})`
      );
    }
    if (!Number.isInteger(args.sequenceLength) || args.sequenceLength <= 0) {
      throw new Error(
        `Station.sampleNext: sequenceLength must be a positive integer (got ${String(
          args.sequenceLength
        )})`
      );
    }
    const body = floatArrayToBody(args.residual);
    const headers: Record<string, string> = {
      'Content-Type': 'application/octet-stream',
      'X-Sequence-Length': String(args.sequenceLength),
      'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
    };
    applyAckermannAdmissionGate(headers, args.ackermannAdmission);
    const res = await this.endpoint.fetch(
      new Request(`${this.baseUrl}/sample-next`, {
        method: 'POST',
        headers,
        body,
      })
    );
    return await readTokenIdResponse(res, '/sample-next');
  }

  /** Single-station native decode: token+position -> greedy next token. */
  async decodeNext(args: {
    tokenId: number;
    position: number;
    requestId?: string;
    ackermannAdmission?: AckermannAdmissionGate;
    /**
     * Optional native terminal-prosody. When supplied, the fat-station
     * applies the SAME finite-horizon cadence selection that `/generate`
     * does, on the forward this call already runs — so the streaming /
     * prefill-window decode loop stays on the one-call-per-token
     * `/decode-next` fast path during the terminal window instead of
     * falling back to the slow `/embed` + `/forward` + `/lm-head` triple
     * that pulls the full logits vector back per step.
     */
    terminalProsody?: {
      emittedTokens: number;
      maxTokens: number;
      tailState: 'open-clause' | 'sentence-closed';
      sentenceCloseTokenIds?: readonly number[];
      stopTokenIds?: readonly number[];
    };
  }): Promise<number> {
    if (!Number.isInteger(args.tokenId) || args.tokenId < 0) {
      throw new Error(
        `Station.decodeNext: tokenId must be a non-negative integer (got ${String(
          args.tokenId
        )})`
      );
    }
    if (!Number.isInteger(args.position) || args.position < 0) {
      throw new Error(
        `Station.decodeNext: position must be a non-negative integer (got ${String(
          args.position
        )})`
      );
    }
    const headers: Record<string, string> = {
      'Content-Type': 'application/octet-stream',
      'X-Token': String(args.tokenId),
      'X-Position': String(args.position),
      'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
    };
    if (args.terminalProsody) {
      const tp = args.terminalProsody;
      headers['X-Prosody-Emitted'] = String(tp.emittedTokens);
      headers['X-Prosody-Max-Tokens'] = String(tp.maxTokens);
      headers['X-Prosody-Tail-State'] = tp.tailState;
      if (tp.sentenceCloseTokenIds && tp.sentenceCloseTokenIds.length > 0) {
        headers['X-Prosody-Sentence-Close'] = Array.from(tp.sentenceCloseTokenIds).join(',');
      }
      if (tp.stopTokenIds && tp.stopTokenIds.length > 0) {
        headers['X-Prosody-Stop'] = Array.from(tp.stopTokenIds).join(',');
      }
    }
    applyAckermannAdmissionGate(headers, args.ackermannAdmission);
    const res = await this.endpoint.fetch(
      new Request(`${this.baseUrl}/decode-next`, {
        method: 'POST',
        headers,
        body: new ArrayBuffer(0),
      })
    );
    return await readTokenIdResponse(res, '/decode-next');
  }

  async generateTokens(args: {
    tokens: Uint32Array;
    maxNewTokens: number;
    requestId?: string;
    terminalProsody?: {
      sentenceCloseTokenIds?: readonly number[];
      stopTokenIds?: readonly number[];
      tailState?: 'open-clause' | 'sentence-closed';
    };
  }): Promise<Uint32Array> {
    if (!Number.isInteger(args.maxNewTokens) || args.maxNewTokens < 0) {
      throw new Error(
        `Station.generateTokens: maxNewTokens must be a non-negative integer (got ${String(
          args.maxNewTokens
        )})`
      );
    }
    const terminalProsody = args.terminalProsody
      ? {
          sentence_close_token_ids: args.terminalProsody.sentenceCloseTokenIds ?? [],
          stop_token_ids: args.terminalProsody.stopTokenIds ?? [],
          tail_state: args.terminalProsody.tailState ?? 'open-clause',
        }
      : undefined;
    const body: Record<string, unknown> = {
      tokens: Array.from(args.tokens),
      max_new_tokens: args.maxNewTokens,
    };
    if (terminalProsody) {
      body['terminal_prosody'] = terminalProsody;
    }
    const res = await this.endpoint.fetch(
      new Request(`${this.baseUrl}/generate`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
        },
        body: JSON.stringify(body),
      })
    );
    return await readGeneratedTokensResponse(res, '/generate');
  }

  /**
   * Sovereign-sieve FOLD: native `/generate` with `return_hidden: true` so a SINGLE
   * forward returns both the generated tokens and `prompt_hidden` (the last prompt
   * token's tail residual = the semantic-sieve key). See fat-station.rs (commit
   * 02618367). `promptHidden` is null if the station omits the field on this knot.
   */
  async generateTokensWithHidden(args: {
    tokens: Uint32Array;
    maxNewTokens: number;
    requestId?: string;
  }): Promise<{ tokens: Uint32Array; promptHidden: Float32Array | null }> {
    if (!Number.isInteger(args.maxNewTokens) || args.maxNewTokens < 0) {
      throw new Error(
        `Station.generateTokensWithHidden: maxNewTokens must be a non-negative integer (got ${String(
          args.maxNewTokens
        )})`
      );
    }
    const body: Record<string, unknown> = {
      tokens: Array.from(args.tokens),
      max_new_tokens: args.maxNewTokens,
      return_hidden: true,
    };
    const res = await this.endpoint.fetch(
      new Request(`${this.baseUrl}/generate`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
        },
        body: JSON.stringify(body),
      })
    );
    return await readGeneratedTokensWithHiddenResponse(res, '/generate');
  }

  /**
   * Prompt-lookup / n-gram speculative-decode VERIFY primitive. POSTs the draft
   * `candidates` to the fat-station `/verify-batch` route, which runs ONE
   * batched forward over the span, accepts the longest argmax-matching prefix,
   * commits `accepted ++ [nextToken]` to KV, and returns the accepted prefix +
   * corrected/continuation token + resulting position. Greedy + LOSSLESS: see
   * IStation.verifyBatch and handle_verify_batch (fat-station.rs).
   */
  async verifyBatch(args: {
    candidates: Uint32Array | readonly number[];
    position?: number;
    promptTokens?: Uint32Array | readonly number[];
    requestId?: string;
  }): Promise<{
    accepted: number[];
    nextToken: number;
    acceptedCount: number;
    candidateCount: number;
    pos: number;
  }> {
    // Empty candidates is valid: a single greedy step (no draft) — the station
    // emits its own argmax next token and commits it.
    const candidates = Array.from(args.candidates);
    for (const c of candidates) {
      if (!Number.isInteger(c) || c < 0 || c > 0xffffffff) {
        throw new Error(
          `Station.verifyBatch: candidate ${String(c)} is not a valid u32`
        );
      }
    }
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
      'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
    };
    // Absolute KV slot of the first candidate. Drives `self.pos` on the
    // fat-station (mirrors `/decode-next` / `/forward`'s X-Position) so the
    // batched candidate forward writes KV at the right slots and never clobbers
    // the warm prompt KV. Omitted only by legacy callers that re-prefill.
    if (args.position !== undefined) {
      if (!Number.isInteger(args.position) || args.position < 0) {
        throw new Error(
          `Station.verifyBatch: position must be a non-negative integer (got ${String(
            args.position
          )})`
        );
      }
      headers['X-Position'] = String(args.position);
    }
    const body: Record<string, unknown> = { candidates };
    if (args.promptTokens && args.promptTokens.length > 0) {
      body['prompt_tokens'] = Array.from(args.promptTokens);
    }
    const res = await this.endpoint.fetch(
      new Request(`${this.baseUrl}/verify-batch`, {
        method: 'POST',
        headers,
        body: JSON.stringify(body),
      })
    );
    return await readVerifyBatchResponse(res, '/verify-batch');
  }

  // Resume-anywhere KV-cache checkpoint for the uniform-KV transformer archs:
  // the native Llama family (/health arch "transformer") AND Gemma (/health arch
  // "gemma") — fat-station.rs ServingPipeline::Llama and ::Gemma, both backed by
  // a single kv_cache::KVCache that get_kv_slab/set_kv_slab serialize. Gate on
  // the /health `kv_session` flag (true for transformer | gemma). SSM/hybrid
  // archs (mamba/rwkv7/hybrid/qwen35moe) have no uniform KV cache: /kv-session
  // returns HTTP 400 for them (this method throws), so they must resume through
  // saveSsmState/loadSsmState instead. A missing route (old build) returns 404 →
  // surfaced as `false`.
  async saveKvSession(args: {
    sessionId: string;
    prefixLength: number;
    requestId?: string;
  }): Promise<boolean> {
    if (!Number.isInteger(args.prefixLength) || args.prefixLength < 0) {
      throw new Error(
        `Station.saveKvSession: prefixLength must be a non-negative integer (got ${String(
          args.prefixLength
        )})`
      );
    }
    const res = await this.endpoint.fetch(
      new Request(`${this.baseUrl}/kv-session/save`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/octet-stream',
          'X-Session-Id': args.sessionId,
          'X-Prefix-Length': String(args.prefixLength),
          'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
        },
        body: new ArrayBuffer(0),
      })
    );
    if (res.status === 404) return false;
    if (!res.ok) {
      throw new Error(
        `Station.saveKvSession: /kv-session/save HTTP ${
          res.status
        }: ${await res.text()}`
      );
    }
    await res.arrayBuffer();
    return true;
  }

  async loadKvSession(args: {
    sessionId: string;
    requestId?: string;
  }): Promise<{ prefixLength: number } | null> {
    const res = await this.endpoint.fetch(
      new Request(`${this.baseUrl}/kv-session/load`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/octet-stream',
          'X-Session-Id': args.sessionId,
          'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
        },
        body: new ArrayBuffer(0),
      })
    );
    if (res.status === 404) return null;
    if (!res.ok) {
      throw new Error(
        `Station.loadKvSession: /kv-session/load HTTP ${
          res.status
        }: ${await res.text()}`
      );
    }
    const json = (await res.json()) as { prefix_length?: unknown };
    const prefixLength =
      typeof json.prefix_length === 'number' &&
      Number.isInteger(json.prefix_length)
        ? json.prefix_length
        : 0;
    return { prefixLength };
  }

  async deleteKvSession(args: {
    sessionId: string;
    requestId?: string;
  }): Promise<boolean> {
    const res = await this.endpoint.fetch(
      new Request(`${this.baseUrl}/kv-session/delete`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/octet-stream',
          'X-Session-Id': args.sessionId,
          'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
        },
        body: new ArrayBuffer(0),
      })
    );
    if (res.status === 404) return false;
    if (!res.ok) {
      throw new Error(
        `Station.deleteKvSession: /kv-session/delete HTTP ${
          res.status
        }: ${await res.text()}`
      );
    }
    await res.arrayBuffer();
    return true;
  }

  // ── O(1) SSM resume-anywhere checkpoint (RWKV-7 + Mamba + hybrids) ───────
  //
  // Mirrors saveKvSession/loadKvSession but for the SSM + SSM-hybrid serving
  // pipelines (fat-station.rs ServingPipeline::Rwkv7, ::Mamba, ::Hybrid, and
  // ::Qwen35Moe). The fat-station /ssm-state/{save,load} routes serialize the
  // COMPLETE recurrent state — RWKV-7: per layer WKV matrix + token-shift +
  // channel-mix + v_first; Mamba/FalconMamba: per layer causal-conv1d buffer +
  // SSM hidden state; Hybrid (Jamba/Falcon-H1) and Qwen3.5-MoE/Qwen3-Next
  // (Gated-DeltaNet): a MIXED blob = O(1) recurrent state for the SSM layers +
  // a KV slab for the sparse attention layers — in the wire layout proven in
  // wasm_bindings_ssm + the ssm-host kill/restore test + vcpu-infer-ssm's
  // MambaKernel resume proof. For the pure-SSM archs the blob is constant-size
  // regardless of position — that is the M5 O(1) headline. Position/last-token
  // stay HOST-owned (same split as the transformer log-rolling path), so they
  // are NOT part of this blob.
  //
  // Gate on the /health `ssm_state` (true for arch rwkv7 | mamba | falcon-mamba
  // | hybrid | qwen35moe) or `alt_arch` flag: a transformer fat-station returns 404 here,
  // which both methods surface as `null` — the gate is enforced server-side by
  // arch. The blob is opaque to the host (raw bytes), so a hybrid's MIXED
  // checkpoint (O(1) mamba state + per-attention-layer KV slab) needs no host
  // change — it round-trips through save/load like any other SSM-state blob.

  /**
   * Pull the live SSM pipeline's (RWKV-7 or Mamba) O(1) recurrent-state
   * checkpoint as raw bytes (octet-stream). Returns null when the endpoint is
   * absent (404) — i.e. this station fronts a transformer knot without the
   * SSM-state plane.
   */
  async saveSsmState(args: {
    requestId?: string;
  }): Promise<Uint8Array | null> {
    const res = await this.endpoint.fetch(
      new Request(`${this.baseUrl}/ssm-state/save`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/octet-stream',
          'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
        },
        body: new ArrayBuffer(0),
      })
    );
    if (res.status === 404) return null;
    if (!res.ok) {
      throw new Error(
        `Station.saveSsmState: /ssm-state/save HTTP ${
          res.status
        }: ${await res.text()}`
      );
    }
    const buf = await res.arrayBuffer();
    return new Uint8Array(buf);
  }

  /**
   * Restore a checkpoint produced by {@link saveSsmState} into the live
   * pipeline state. Restoring into a FRESH fat-station serving the SAME knot
   * reproduces the exact continuation (resume-anywhere). Returns null when the
   * endpoint is absent (404).
   */
  async loadSsmState(args: {
    state: Uint8Array | ArrayBuffer;
    requestId?: string;
  }): Promise<{ restoredBytes: number; numLayers: number } | null> {
    const bytes =
      args.state instanceof Uint8Array
        ? args.state
        : new Uint8Array(args.state);
    if (bytes.byteLength < 4) {
      throw new Error(
        `Station.loadSsmState: checkpoint too short (${bytes.byteLength} bytes)`
      );
    }
    const body = bytes.buffer.slice(
      bytes.byteOffset,
      bytes.byteOffset + bytes.byteLength
    ) as ArrayBuffer;
    const res = await this.endpoint.fetch(
      new Request(`${this.baseUrl}/ssm-state/load`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/octet-stream',
          'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
        },
        body,
      })
    );
    if (res.status === 404) return null;
    if (!res.ok) {
      throw new Error(
        `Station.loadSsmState: /ssm-state/load HTTP ${
          res.status
        }: ${await res.text()}`
      );
    }
    const json = (await res.json()) as {
      restored_bytes?: unknown;
      num_layers?: unknown;
    };
    const restoredBytes =
      typeof json.restored_bytes === 'number' &&
      Number.isInteger(json.restored_bytes)
        ? json.restored_bytes
        : bytes.byteLength;
    const numLayers =
      typeof json.num_layers === 'number' && Number.isInteger(json.num_layers)
        ? json.num_layers
        : 0;
    return { restoredBytes, numLayers };
  }

  // ── Pneuma (Whisper STT) endpoints ───────────────────────────────────────
  //
  // The methods below front the encoder/decoder surface that
  // {@link import('./pipeline_pneuma').PneumaPipeline} dispatches against.
  // They are additive to the LLM endpoints above — each maps to a distinct
  // path on the remote Worker:
  //
  //   POST /embed                     embedAudio
  //   POST /forward-range             forwardEncoderRange
  //   POST /finalize                  finalizeEncoder
  //   POST /encode-streaming          encodeStreamingChunk        (Wave 3)
  //   POST /load-audio-context        loadAudioContext
  //   POST /decode-step               decodeStep
  //
  // Wire format mirrors the LLM surface: raw Float32Array body for the
  // residual / encoder output, ASCII headers for metadata. Keeping the
  // octet-stream shape lets the same Worker-side reader code path serve
  // both LLMs and Whisper.

  /**
   * Pneuma encoder entry: ship a log-mel spectrogram to the entry encoder
   * station, return the post-conv (and, for single-station deploys, the
   * post-encoder) residual.
   *
   * `mel`: row-major `(n_mels, n_frames)` flat Float32Array. The remote
   * Worker is the authority on the expected shape (typically
   * `80 * 3000 = 240_000` for whisper-tiny).
   */
  async embedAudio(args: {
    mel: Float32Array;
    requestId?: string;
  }): Promise<Float32Array> {
    const body = floatArrayToBody(args.mel);
    const headers: Record<string, string> = {
      'Content-Type': 'application/octet-stream',
      'X-Layer-Range': `${this.startLayer}-${this.endLayer}`,
      'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
    };
    const res = await this.endpoint.fetch(
      new Request(`${this.baseUrl}/embed`, {
        method: 'POST',
        headers,
        body,
      })
    );
    return await readFloatArrayResponse(res, '/embed');
  }

  /**
   * Pneuma encoder middle hop: walk the residual through this station's
   * pinned encoder layer range. Stateless across calls (no KV cache).
   */
  async forwardEncoderRange(args: {
    residual: Float32Array;
    requestId?: string;
  }): Promise<Float32Array> {
    const body = floatArrayToBody(args.residual);
    const headers: Record<string, string> = {
      'Content-Type': 'application/octet-stream',
      'X-Layer-Range': `${this.startLayer}-${this.endLayer}`,
      'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
    };
    const res = await this.endpoint.fetch(
      new Request(`${this.baseUrl}/forward-range`, {
        method: 'POST',
        headers,
        body,
      })
    );
    return await readFloatArrayResponse(res, '/forward-range');
  }

  /**
   * Pneuma encoder exit: apply the final LayerNorm to the residual and
   * return the `(n_audio_ctx, hidden)` encoder output the decoder will
   * cross-attend to.
   */
  async finalizeEncoder(args: {
    residual: Float32Array;
    requestId?: string;
  }): Promise<Float32Array> {
    const body = floatArrayToBody(args.residual);
    const headers: Record<string, string> = {
      'Content-Type': 'application/octet-stream',
      'X-Layer-Range': `${this.startLayer}-${this.endLayer}`,
      'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
    };
    const res = await this.endpoint.fetch(
      new Request(`${this.baseUrl}/finalize`, {
        method: 'POST',
        headers,
        body,
      })
    );
    return await readFloatArrayResponse(res, '/finalize');
  }

  /**
   * Pneuma streaming encoder: feed one mel chunk to the encoder station
   * and pull back the partial encoder output reflecting all chunks fed so
   * far in the current utterance.
   *
   * The Worker is expected to:
   *   1. Call `WhisperEncoder::forward_streaming_chunk(chunk_mel, chunkIndex)`
   *      to append the chunk to its 30 s mel ring.
   *   2. Call `WhisperEncoder::forward_streaming_finalize()` to compute the
   *      partial encoder output over the populated ring (zero-padded out
   *      to 30 s for the unfilled tail).
   *   3. Return the resulting `(n_audio_ctx, hidden)` Float32Array.
   *
   * `chunkIndex == 0` signals the start of a new utterance — the Worker
   * MUST call `WhisperEncoder::reset_streaming` before step (1) so a fresh
   * mel ring is used.
   *
   * Because the v1 implementation re-runs the full encoder on every chunk,
   * the time per chunk is independent of `chunkIndex`. Real chunked
   * attention with cached K/V is a Phase 3+ optimisation.
   *
   * `melChunk`: row-major `(n_mels, chunk_n_frames)` Float32Array. The
   * caller picks the chunk size — 1 s of audio at 100 Hz = 100 frames is
   * a reasonable default; sub-second VAD-driven chunks are also fine.
   */
  async encodeStreamingChunk(args: {
    melChunk: Float32Array;
    chunkIndex: number;
    requestId?: string;
  }): Promise<Float32Array> {
    if (!Number.isInteger(args.chunkIndex) || args.chunkIndex < 0) {
      throw new Error(
        `Station.encodeStreamingChunk: chunkIndex must be a non-negative integer (got ${String(
          args.chunkIndex
        )})`
      );
    }
    const body = floatArrayToBody(args.melChunk);
    const headers: Record<string, string> = {
      'Content-Type': 'application/octet-stream',
      'X-Chunk-Index': String(args.chunkIndex),
      'X-Layer-Range': `${this.startLayer}-${this.endLayer}`,
      'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
    };
    const res = await this.endpoint.fetch(
      new Request(`${this.baseUrl}/encode-streaming`, {
        method: 'POST',
        headers,
        body,
      })
    );
    return await readFloatArrayResponse(res, '/encode-streaming');
  }

  /**
   * Pneuma decoder priming: ship the encoder output to a decoder station
   * so it can populate its frozen cross-attention K/V cache (and clear
   * its growing self-attention cache) before the autoregressive loop
   * starts. Called once per utterance per decoder station.
   *
   * The remote Worker must invoke `WhisperDecoder::prepare_audio_context`
   * + `WhisperDecoder::reset` in its handler. No payload comes back; we
   * still return a `Promise<void>` for await-discipline at the call site.
   */
  async loadAudioContext(args: {
    encoderOutput: Float32Array;
    requestId?: string;
  }): Promise<void> {
    const body = floatArrayToBody(args.encoderOutput);
    const headers: Record<string, string> = {
      'Content-Type': 'application/octet-stream',
      'X-Layer-Range': `${this.startLayer}-${this.endLayer}`,
      'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
    };
    const res = await this.endpoint.fetch(
      new Request(`${this.baseUrl}/load-audio-context`, {
        method: 'POST',
        headers,
        body,
      })
    );
    if (!res.ok) {
      let detail = '';
      try {
        detail = await res.text();
      } catch {
        // ignore
      }
      throw new Error(
        `Station /load-audio-context: HTTP ${res.status} ${res.statusText}${
          detail ? ` -- ${detail}` : ''
        }`
      );
    }
    // Drain body for connection-reuse hygiene — handler may legitimately
    // return an empty body (no logits to ship back).
    try {
      await res.arrayBuffer();
    } catch {
      // ignore
    }
  }

  /**
   * Pneuma decoder step: run one autoregressive token through the decoder
   * station and return the next-token logits.
   *
   * `token`: previous token id (BOS for the first step).
   * `pos`: 0-based position in the generated sequence.
   * `resetSelfKV`: optional. When `true`, the Worker MUST call
   *   `WhisperDecoder::reset` before the step to wipe the self-attention
   *   cache. Used by `transcribeStreaming` to re-decode against a fresh
   *   partial encoder output without leaking state from the previous
   *   pass. Defaults to `false` (the standard autoregressive contract).
   */
  async decodeStep(args: {
    token: number;
    pos: number;
    resetSelfKV?: boolean;
    requestId?: string;
  }): Promise<Float32Array> {
    if (!Number.isInteger(args.token) || args.token < 0) {
      throw new Error(
        `Station.decodeStep: token must be a non-negative integer (got ${String(
          args.token
        )})`
      );
    }
    if (!Number.isInteger(args.pos) || args.pos < 0) {
      throw new Error(
        `Station.decodeStep: pos must be a non-negative integer (got ${String(
          args.pos
        )})`
      );
    }
    // Empty body — token + pos are header-borne. Saves a serialisation
    // round trip vs. JSON for what is a very hot per-token path.
    const headers: Record<string, string> = {
      'Content-Type': 'application/octet-stream',
      'X-Token': String(args.token),
      'X-Position': String(args.pos),
      'X-Layer-Range': `${this.startLayer}-${this.endLayer}`,
      'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
    };
    if (args.resetSelfKV) {
      headers['X-Reset-Self-KV'] = '1';
    }
    const res = await this.endpoint.fetch(
      new Request(`${this.baseUrl}/decode-step`, {
        method: 'POST',
        headers,
        body: new ArrayBuffer(0),
      })
    );
    return await readFloatArrayResponse(res, '/decode-step');
  }

  /**
   * TQFT Cobordism: vocabulary-sharded embedding lookup.
   *
   * Sends token IDs to a cobordism worker that owns vocab range
   * [vocabStart, vocabEnd). Returns partial embeddings: non-zero for
   * tokens in range, zero for out-of-range. The coordinator merges
   * via elementwise addition (Frobenius μ: A ⊗ A → A).
   *
   * See: FORMAL_LEDGER.md § Atiyah-Segal TQFT Functor
   */
  async cobordismEmbed(args: {
    tokens: Uint32Array;
    startPos: number;
    vocabStart: number;
    vocabEnd: number;
    requestId?: string;
  }): Promise<Float32Array> {
    const body = uintArrayToBody(args.tokens);
    const headers: Record<string, string> = {
      'Content-Type': 'application/octet-stream',
      'X-Position': String(args.startPos),
      'X-Token-Count': String(args.tokens.length),
      'X-Vocab-Start': String(args.vocabStart),
      'X-Vocab-End': String(args.vocabEnd),
      'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
    };
    const res = await this.endpoint.fetch(
      new Request(`${this.baseUrl}/cobordism-embed`, {
        method: 'POST',
        headers,
        body,
      })
    );
    return await readFloatArrayResponse(res, '/cobordism-embed');
  }

  /**
   * TQFT Cobordism: vocabulary-sharded lm_head (comultiplication).
   *
   * Sends the final residual to a cobordism worker that computes partial
   * logits over [vocabStart, vocabEnd). The coordinator concatenates
   * partial logits from all shards to reconstruct the full logits vector.
   *
   * This is the comultiplication (Δ: A → A ⊗ A), dual to cobordismEmbed.
   */
  async cobordismLmHead(args: {
    residual: Float32Array;
    sequenceLength: number;
    vocabStart: number;
    vocabEnd: number;
    requestId?: string;
  }): Promise<Float32Array> {
    const body = floatArrayToBody(args.residual);
    const headers: Record<string, string> = {
      'Content-Type': 'application/octet-stream',
      'X-Sequence-Length': String(args.sequenceLength),
      'X-Vocab-Start': String(args.vocabStart),
      'X-Vocab-End': String(args.vocabEnd),
      'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
    };
    const res = await this.endpoint.fetch(
      new Request(`${this.baseUrl}/cobordism-lm-head`, {
        method: 'POST',
        headers,
        body,
      })
    );
    return await readFloatArrayResponse(res, '/cobordism-lm-head');
  }

  // ── Amplituhedron admin path ───────────────────────────────────────────
  //
  // The Amplituhedron volume cache (AMPLITUHEDRON_PLAN.md §4) is a
  // bitwise-binary admin path on the same HTTPS endpoint. Pipeline fans
  // `/amplituhedron/replay` across every station to decide whether the
  // prefill can be skipped; on miss it fires `/amplituhedron/capture` per
  // station. Both calls MUST be best-effort — any failure degrades to a
  // miss + normal prefill and MUST NOT throw on the hot path. The wire
  // format is the bitwise codec (see ./amplituhedron-wire), ~7x smaller
  // than the prior JSON shape on the residual payload.

  /**
   * Amplituhedron volume-cache replay. POSTs the key tuple as JSON to
   * `/amplituhedron/replay` and reads the bitwise-binary response envelope
   * via `decodeReplayResponse`. Returns `{ hit: false }` on 204 No Content
   * (miss), the decoded `{ hit, tailResidual }` on 200, and
   * `{ hit: false }` defensively on any other status / decode error so a
   * flaky admin route can never poison the coordinator's hot path.
   */
  async amplituhedronReplay(args: {
    prefixHash: string;
    prefixLen: number;
    layerLo: number;
    layerHi: number;
    requestId?: string;
  }): Promise<{
    hit: boolean;
    tailResidual?: Float32Array;
    tailFingerprint?: string;
    prefixResidual?: Float32Array;
    prefixFingerprint?: string;
  }> {
    const body = JSON.stringify({
      prefix_hash: args.prefixHash,
      prefix_len: args.prefixLen,
      layer_lo: args.layerLo,
      layer_hi: args.layerHi,
      ...(args.requestId !== undefined ? { request_id: args.requestId } : {}),
    });
    let res: Response;
    try {
      res = await this.endpoint.fetch(
        new Request(`${this.baseUrl}/amplituhedron/replay`, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            Accept: AMPLITUHEDRON_CONTENT_TYPE,
            'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
          },
          body,
        })
      );
    } catch {
      // Network errors fall back to MISS — the coordinator handles this by
      // running a normal prefill. Never throw on the replay path.
      return { hit: false };
    }
    if (res.status === 204) return { hit: false };
    if (res.status === 200) {
      try {
        const buf = await res.arrayBuffer();
        return decodeReplayResponse(new Uint8Array(buf));
      } catch {
        return { hit: false };
      }
    }
    return { hit: false };
  }

  /**
   * Fire-and-forget capture POST. The coordinator never awaits this (see
   * Pipeline.scheduleAmplituhedronCapture → pipeline.ts:964), so we
   * swallow every error here: a failed capture just means the next
   * same-prefix call misses and re-runs prefill — a perf regression,
   * not a correctness bug. Body is encoded via the bitwise codec.
   */
  async amplituhedronCapture(args: {
    prefixHash: string;
    prefixLen: number;
    layerLo: number;
    layerHi: number;
    tailResidual?: Float32Array;
    requestId?: string;
  }): Promise<void> {
    // Contract gap: IStation declares tailResidual optional (callers like
    // pipeline.ts may omit it on misses where the tail wasn't materialized),
    // but the wire codec requires a Float32Array. Resolve at the impl layer:
    // no residual = nothing to capture = fire-and-forget noop.
    if (!args.tailResidual) {
      return;
    }
    let body: Uint8Array;
    try {
      body = encodeCaptureRequest({
        prefixHash: args.prefixHash,
        prefixLen: args.prefixLen,
        layerLo: args.layerLo,
        layerHi: args.layerHi,
        tailResidual: args.tailResidual,
      });
    } catch {
      // Encoder failure is fatal to capture; just drop the call.
      return;
    }
    try {
      await this.endpoint.fetch(
        new Request(`${this.baseUrl}/amplituhedron/capture`, {
          method: 'POST',
          headers: {
            'Content-Type': AMPLITUHEDRON_CONTENT_TYPE,
            'X-Request-Id': args.requestId ?? `req-${Date.now()}`,
          },
          body: body.buffer.slice(
            body.byteOffset,
            body.byteOffset + body.byteLength,
          ) as ArrayBuffer,
        })
      );
    } catch {
      // Capture is best-effort. A failure here just means the next call
      // misses and re-runs prefill. Never throw.
    }
  }
}

// ── helpers ────────────────────────────────────────────────────────────────

/** Copy a Float32Array view into a fresh ArrayBuffer suitable for Request.body. */
function floatArrayToBody(arr: Float32Array): ArrayBuffer {
  const buf = new ArrayBuffer(arr.byteLength);
  new Uint8Array(buf).set(
    new Uint8Array(arr.buffer, arr.byteOffset, arr.byteLength)
  );
  return buf;
}

function uintArrayToBody(arr: Uint32Array): ArrayBuffer {
  const buf = new ArrayBuffer(arr.byteLength);
  new Uint8Array(buf).set(
    new Uint8Array(arr.buffer, arr.byteOffset, arr.byteLength)
  );
  return buf;
}

async function readFloatArrayResponse(
  res: Response,
  label: string
): Promise<Float32Array> {
  if (!res.ok) {
    let detail = '';
    try {
      detail = await res.text();
    } catch {
      // ignore
    }
    throw new Error(
      `Station ${label}: HTTP ${res.status} ${res.statusText}${
        detail ? ` -- ${detail}` : ''
      }`
    );
  }
  const buf = await res.arrayBuffer();
  if (buf.byteLength % 4 !== 0) {
    throw new Error(
      `Station ${label}: response byteLength ${buf.byteLength} is not a multiple of 4 (Float32)`
    );
  }
  return new Float32Array(buf);
}

async function readTokenIdResponse(
  res: Response,
  label: string
): Promise<number> {
  if (!res.ok) {
    let detail = '';
    try {
      detail = await res.text();
    } catch {
      // ignore
    }
    throw new Error(
      `Station ${label}: HTTP ${res.status} ${res.statusText}${
        detail ? ` -- ${detail}` : ''
      }`
    );
  }
  const payload = (await res.json()) as { token_id?: unknown };
  if (
    typeof payload.token_id !== 'number' ||
    !Number.isFinite(payload.token_id) ||
    !Number.isInteger(payload.token_id) ||
    payload.token_id < 0
  ) {
    throw new Error(`Station ${label}: invalid token_id response`);
  }
  return payload.token_id;
}

async function readGeneratedTokensResponse(
  res: Response,
  label: string
): Promise<Uint32Array> {
  if (!res.ok) {
    let detail = '';
    try {
      detail = await res.text();
    } catch {
      // ignore
    }
    throw new Error(
      `Station ${label}: HTTP ${res.status} ${res.statusText}${
        detail ? ` -- ${detail}` : ''
      }`
    );
  }
  const payload = (await res.json()) as { generated_tokens?: unknown };
  if (!Array.isArray(payload.generated_tokens)) {
    throw new Error(`Station ${label}: invalid generated_tokens response`);
  }
  const out = new Uint32Array(payload.generated_tokens.length);
  for (let i = 0; i < payload.generated_tokens.length; i++) {
    const token = payload.generated_tokens[i];
    if (
      typeof token !== 'number' ||
      !Number.isFinite(token) ||
      !Number.isInteger(token) ||
      token < 0 ||
      token > 0xffffffff
    ) {
      throw new Error(`Station ${label}: invalid generated token at index ${i}`);
    }
    out[i] = token;
  }
  return out;
}

/**
 * Read a `/generate` response that was issued with `return_hidden: true`. Same
 * `generated_tokens` decode as {@link readGeneratedTokensResponse}, plus the
 * optional `prompt_hidden` f32 array (the sovereign-sieve semantic key). A missing
 * or malformed `prompt_hidden` yields `promptHidden: null` (capability miss) rather
 * than throwing — the sieve lane must degrade to a normal decode, never fail.
 */
async function readGeneratedTokensWithHiddenResponse(
  res: Response,
  label: string
): Promise<{ tokens: Uint32Array; promptHidden: Float32Array | null }> {
  if (!res.ok) {
    let detail = '';
    try {
      detail = await res.text();
    } catch {
      // ignore
    }
    throw new Error(
      `Station ${label}: HTTP ${res.status} ${res.statusText}${
        detail ? ` -- ${detail}` : ''
      }`
    );
  }
  const payload = (await res.json()) as {
    generated_tokens?: unknown;
    prompt_hidden?: unknown;
  };
  if (!Array.isArray(payload.generated_tokens)) {
    throw new Error(`Station ${label}: invalid generated_tokens response`);
  }
  const out = new Uint32Array(payload.generated_tokens.length);
  for (let i = 0; i < payload.generated_tokens.length; i++) {
    const token = payload.generated_tokens[i];
    if (
      typeof token !== 'number' ||
      !Number.isFinite(token) ||
      !Number.isInteger(token) ||
      token < 0 ||
      token > 0xffffffff
    ) {
      throw new Error(`Station ${label}: invalid generated token at index ${i}`);
    }
    out[i] = token;
  }
  let promptHidden: Float32Array | null = null;
  if (Array.isArray(payload.prompt_hidden) && payload.prompt_hidden.length > 0) {
    const h = new Float32Array(payload.prompt_hidden.length);
    let valid = true;
    for (let i = 0; i < payload.prompt_hidden.length; i++) {
      const v = payload.prompt_hidden[i];
      if (typeof v !== 'number' || !Number.isFinite(v)) {
        valid = false;
        break;
      }
      h[i] = v;
    }
    if (valid) promptHidden = h;
  }
  return { tokens: out, promptHidden };
}

async function readVerifyBatchResponse(
  res: Response,
  label: string
): Promise<{
  accepted: number[];
  nextToken: number;
  acceptedCount: number;
  candidateCount: number;
  pos: number;
}> {
  if (!res.ok) {
    let detail = '';
    try {
      detail = await res.text();
    } catch {
      // ignore
    }
    throw new Error(
      `Station ${label}: HTTP ${res.status} ${res.statusText}${
        detail ? ` -- ${detail}` : ''
      }`
    );
  }
  const payload = (await res.json()) as {
    accepted?: unknown;
    next_token?: unknown;
    accepted_count?: unknown;
    candidate_count?: unknown;
    pos?: unknown;
  };
  if (!Array.isArray(payload.accepted)) {
    throw new Error(`Station ${label}: invalid accepted array`);
  }
  const accepted: number[] = [];
  for (let i = 0; i < payload.accepted.length; i++) {
    const t = payload.accepted[i];
    if (
      typeof t !== 'number' ||
      !Number.isInteger(t) ||
      t < 0 ||
      t > 0xffffffff
    ) {
      throw new Error(`Station ${label}: invalid accepted token at index ${i}`);
    }
    accepted.push(t);
  }
  const isU32 = (v: unknown): v is number =>
    typeof v === 'number' && Number.isInteger(v) && v >= 0;
  if (!isU32(payload.next_token)) {
    throw new Error(`Station ${label}: invalid next_token`);
  }
  if (!isU32(payload.pos)) {
    throw new Error(`Station ${label}: invalid pos`);
  }
  const acceptedCount = isU32(payload.accepted_count)
    ? payload.accepted_count
    : accepted.length;
  const candidateCount = isU32(payload.candidate_count)
    ? payload.candidate_count
    : 0;
  return {
    accepted,
    nextToken: payload.next_token,
    acceptedCount,
    candidateCount,
    pos: payload.pos,
  };
}

function parseTeleportationHeaders(
  headers: Headers
): { layerDeficits: number[]; rounds: number } | undefined {
  const deficitsStr = headers.get('X-Teleport-Deficits');
  const roundsStr = headers.get('X-Teleport-Rounds');
  if (!deficitsStr || !roundsStr) return undefined;

  try {
    const layerDeficits = deficitsStr
      .split(',')
      .map((v) => parseInt(v.trim(), 10));
    const rounds = parseInt(roundsStr, 10);
    if (layerDeficits.every((v) => !isNaN(v)) && !isNaN(rounds)) {
      return { layerDeficits, rounds };
    }
  } catch {
    // ignore malformed headers
  }
  return undefined;
}
