// @a0n/.ignored_gnosis/distributed-inference-host/src/style-public.ts
export {
  styleTransferAnyImage,
  encodeStyledAsPNG,
} from './style-transfer-any-image';

export {
  runClassicalStyleTransfer,
} from './style-transfer-classical';

export {
  runAdaINAdapter,
  type AdaINAdapterBackend,
  type AdaINAdapterRequest,
} from './style-transfer-adain-adapter';

export {
  forkRaceFoldStyleTransfer,
} from './forkRaceFoldStyleTransfer';

export {
  compositeStyleTransferFoldScore,
  STYLE_TRANSFER_MESH_GATE_SCHEMA_VERSION,
  STYLE_TRANSFER_MESH_SIX_CHANNEL_IDS,
} from './style-transfer-mesh-gate';

export {
  resolveTopologyAlphaForStyle,
  stubTopologyAlphaFromContentRgb,
  STYLE_TRANSFER_TOPOLOGY_GLUE_SCHEMA_VERSION,
} from './style-transfer-topology-glue';

export type {
  TopologyAlphaStyleBackend,
  TopologyAlphaStyleRequest,
  TopologyAlphaStyleResult,
} from './style-transfer-topology-glue';

export {
  topologyAlphaStyleBackendFromImagingWasm,
  tryWasmMeshDepthConfidence,
} from './style-transfer-imaging-wasm';

export type {
  StyleTransferImagingWasmSurface,
  WasmMeshDepthStyleResultLike,
  WasmTopologyMaskResultLike,
} from './style-transfer-imaging-wasm';

export type {
  StyleTransferAnyImageOptions,
  StyleTransferAnyImageResult,
  StyleTransferCandidateReport,
  StyleTransferChannelStats,
  StyleTransferEnsembleArgs,
  StyleTransferEnsembleConfig,
  StyleTransferEnsembleResult,
  StyleTransferImageInput,
  StyleTransferMeshAdmissionChannel,
  StyleTransferMeshFoldMetrics,
  StyleTransferMeshProvenance,
  StyleTransferMethod,
} from './style-transfer-types';
