// @a0n/.ignored_gnosis/distributed-inference-host/src/style-transfer-classical.ts
import {
  buildEffectiveGateMap,
  buildSixChannelAdmissionRows,
  computeMeshGateDiagnostics,
  STYLE_TRANSFER_MESH_GATE_SCHEMA_VERSION,
} from './style-transfer-mesh-gate';
import type {
  StyleTransferAnyImageResult,
  StyleTransferChannelStats,
  StyleTransferMeshProvenance,
} from './style-transfer-types';

interface ClassicalStyleTransferArgs {
  content: Float32Array;
  style: Float32Array;
  width: number;
  height: number;
  channels: 1 | 3;
  strength: number;
  structurePreservation: number;
  alphaWave: {
    enabled: boolean;
    edgeThreshold: number;
    edgeBoost: number;
    frequency: number;
    phase: number;
  };
  sharpen: boolean;
  /** When set, per-pixel style strength is multiplied by combined gate maps. */
  meshGate?: {
    topologyAlpha: Float32Array;
    depthConfidence?: Float32Array | null;
    photometricVeto?: Float32Array | null;
    topologySource: 'wasm' | 'stub';
    topologyProvenanceFragment: Readonly<Record<string, unknown>>;
    depthProvenanceFragment: Readonly<Record<string, unknown>>;
  };
}

/**
 * Handles the gnosis run Classical Style Transfer workflow.
 */
export function runClassicalStyleTransfer(
  args: ClassicalStyleTransferArgs,
): StyleTransferAnyImageResult {
  const {
    content,
    style,
    width,
    height,
    channels,
    strength,
    structurePreservation,
    alphaWave,
    sharpen,
    meshGate,
  } = args;
  const n = width * height;
  const alphaWaveState = buildAlphaWaveState(style, width, height, channels, alphaWave);
  const contentStats = channelStats(content, width, height, channels);
  const styleStats = channelStats(alphaWaveState.wavedStyle, width, height, channels);
  const stylized = new Float32Array(content.length);
  const effectiveGate =
    meshGate !== undefined
      ? buildEffectiveGateMap({
          pixelCount: n,
          alphaWave: alphaWaveState.alphaMap,
          topologyAlpha: meshGate.topologyAlpha,
          depthConfidence: meshGate.depthConfidence ?? null,
          photometricVeto: meshGate.photometricVeto ?? null,
        })
      : null;
  const gateDiag =
    effectiveGate !== null && meshGate !== undefined
      ? computeMeshGateDiagnostics(
          alphaWaveState.alphaMap,
          effectiveGate,
          meshGate.topologyAlpha,
          meshGate.depthConfidence ?? null,
          meshGate.photometricVeto ?? null,
        )
      : null;

  for (let c = 0; c < channels; c++) {
    const cMean = contentStats[c]!.mean;
    const cStd = contentStats[c]!.stdDev;
    const sMean = styleStats[c]!.mean;
    const sStd = styleStats[c]!.stdDev;
    const safeStd = cStd < 1e-6 ? 1 : cStd;
    const gain = sStd / safeStd;
    for (let i = 0; i < n; i++) {
      const idx = i * channels + c;
      const v = content[idx]!;
      const mapped = (v - cMean) * gain + sMean;
      // Style is first passed through alpha-wave; then transposed onto target.
      const waveAlpha = alphaWaveState.alphaMap[i]!;
      const g = effectiveGate !== null ? effectiveGate[i]! : waveAlpha;
      const blended = v + strength * g * (mapped - v);
      stylized[idx] = clamp01(blended);
    }
  }

  const structureMixed = structurePreservationMix(
    content,
    stylized,
    width,
    height,
    channels,
    structurePreservation,
  );
  const post = sharpen ? unsharpMask(structureMixed, width, height, channels) : structureMixed;
  const outputStats = channelStats(post, width, height, channels);
  const styleDistance = statsDistance(styleStats, outputStats);

  let meshProvenance: StyleTransferMeshProvenance | undefined;
  if (gateDiag !== null && meshGate !== undefined) {
    meshProvenance = {
      schema_version: STYLE_TRANSFER_MESH_GATE_SCHEMA_VERSION,
      topology_source: meshGate.topologySource,
      topology_provenance_fragment: meshGate.topologyProvenanceFragment,
      depth_provenance_fragment: meshGate.depthProvenanceFragment,
      channels: buildSixChannelAdmissionRows(gateDiag),
      mean_effective_gate: gateDiag.meanEffectiveGate,
      optional_depth_used:
        meshGate.depthConfidence !== null &&
        meshGate.depthConfidence !== undefined &&
        meshGate.depthConfidence.length === n,
    };
  }

  return {
    stylized: post,
    width,
    height,
    channels,
    methodUsed: 'classical',
    styleDistance,
    strengthUsed: strength,
    structurePreservationUsed: structurePreservation,
    alphaWaveUsed: {
      enabled: alphaWave.enabled,
      edgeThreshold: alphaWave.edgeThreshold,
      edgeBoost: alphaWave.edgeBoost,
      frequency: alphaWave.frequency,
      phase: alphaWave.phase,
      meanAlpha: alphaWaveState.meanAlpha,
    },
    contentStats,
    styleStats,
    outputStats,
    meshProvenance,
    meshFoldMetrics:
      gateDiag !== null
        ? {
            mean_topology_alpha: gateDiag.meanTopologyAlpha,
            mean_depth_confidence: gateDiag.meanDepthConfidence,
            mean_effective_gate: gateDiag.meanEffectiveGate,
            gate_violation_score: gateDiag.gateViolationScore,
          }
        : undefined,
  };
}

function channelStats(
  data: Float32Array,
  width: number,
  height: number,
  channels: 1 | 3,
): StyleTransferChannelStats[] {
  const n = width * height;
  const out: StyleTransferChannelStats[] = [];
  for (let c = 0; c < channels; c++) {
    let sum = 0;
    for (let i = 0; i < n; i++) {
      sum += data[i * channels + c]!;
    }
    const mean = sum / n;
    let varSum = 0;
    for (let i = 0; i < n; i++) {
      const d = data[i * channels + c]! - mean;
      varSum += d * d;
    }
    out.push({ mean, stdDev: Math.sqrt(varSum / n) });
  }
  return out;
}

function statsDistance(
  target: ReadonlyArray<StyleTransferChannelStats>,
  actual: ReadonlyArray<StyleTransferChannelStats>,
): number {
  const n = Math.min(target.length, actual.length);
  let acc = 0;
  for (let i = 0; i < n; i++) {
    const dMean = target[i]!.mean - actual[i]!.mean;
    const dStd = target[i]!.stdDev - actual[i]!.stdDev;
    acc += dMean * dMean + dStd * dStd;
  }
  return Math.sqrt(acc / Math.max(1, n));
}

function structurePreservationMix(
  content: Float32Array,
  stylized: Float32Array,
  width: number,
  height: number,
  channels: 1 | 3,
  structurePreservation: number,
): Float32Array {
  if (structurePreservation <= 0) return stylized;
  if (structurePreservation >= 1) return new Float32Array(content);
  const contentEdge = edgeMagnitude(content, width, height, channels);
  const styleEdge = edgeMagnitude(stylized, width, height, channels);
  const out = new Float32Array(stylized.length);
  for (let i = 0; i < out.length; i++) {
    const edgeMix = structurePreservation * contentEdge[i]! + (1 - structurePreservation) * styleEdge[i]!;
    const raw = stylized[i]! + 0.35 * (edgeMix - styleEdge[i]!);
    out[i] = clamp01(raw);
  }
  return out;
}

function buildAlphaWaveState(
  style: Float32Array,
  width: number,
  height: number,
  channels: 1 | 3,
  wave: { enabled: boolean; edgeThreshold: number; edgeBoost: number; frequency: number; phase: number },
): { alphaMap: Float32Array; meanAlpha: number; wavedStyle: Float32Array } {
  const n = width * height;
  if (!wave.enabled) {
    const alphaMap = new Float32Array(n);
    alphaMap.fill(1);
    return { alphaMap, meanAlpha: 1, wavedStyle: new Float32Array(style) };
  }
  const styleLuma = channels === 1 ? style : rgbToLuminance(style, width, height);
  const edge = edgeMagnitude(styleLuma, width, height, 1);
  const alphaMap = new Float32Array(n);
  const wavedStyle = new Float32Array(style.length);
  let alphaSum = 0;
  const twoPiF = 2 * Math.PI * wave.frequency;
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      const p = y * width + x;
      const e = edge[p]!;
      const edgeNorm = e <= wave.edgeThreshold ? 0 : clamp01((e - wave.edgeThreshold) / (1 - wave.edgeThreshold));
      const oscillation = 0.5 + 0.5 * Math.sin(twoPiF * (x / width + y / height) + wave.phase);
      const alpha = clamp01(((1 - wave.edgeBoost) + wave.edgeBoost * edgeNorm) * oscillation);
      alphaMap[p] = alpha;
      alphaSum += alpha;
      for (let c = 0; c < channels; c++) {
        const idx = p * channels + c;
        wavedStyle[idx] = clamp01(style[idx]! * alpha);
      }
    }
  }
  return { alphaMap, meanAlpha: alphaSum / n, wavedStyle };
}

function edgeMagnitude(
  data: Float32Array,
  width: number,
  height: number,
  channels: 1 | 3,
): Float32Array {
  const out = new Float32Array(data.length);
  const gx = [-1, 0, 1, -2, 0, 2, -1, 0, 1];
  const gy = [1, 2, 1, 0, 0, 0, -1, -2, -1];
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      for (let c = 0; c < channels; c++) {
        let sx = 0;
        let sy = 0;
        let k = 0;
        for (let oy = -1; oy <= 1; oy++) {
          for (let ox = -1; ox <= 1; ox++, k++) {
            const px = clampInt(x + ox, 0, width - 1);
            const py = clampInt(y + oy, 0, height - 1);
            const idx = (py * width + px) * channels + c;
            const v = data[idx]!;
            sx += gx[k]! * v;
            sy += gy[k]! * v;
          }
        }
        const dst = (y * width + x) * channels + c;
        out[dst] = Math.sqrt(sx * sx + sy * sy);
      }
    }
  }
  return out;
}

function rgbToLuminance(data: Float32Array, width: number, height: number): Float32Array {
  const out = new Float32Array(width * height);
  for (let i = 0; i < width * height; i++) {
    const r = data[i * 3]!;
    const g = data[i * 3 + 1]!;
    const b = data[i * 3 + 2]!;
    out[i] = 0.2126 * r + 0.7152 * g + 0.0722 * b;
  }
  return out;
}

function unsharpMask(
  data: Float32Array,
  width: number,
  height: number,
  channels: 1 | 3,
): Float32Array {
  const blurred = boxBlur3x3(data, width, height, channels);
  const out = new Float32Array(data.length);
  const amount = 0.45;
  for (let i = 0; i < data.length; i++) {
    out[i] = clamp01(data[i]! + amount * (data[i]! - blurred[i]!));
  }
  return out;
}

function boxBlur3x3(
  data: Float32Array,
  width: number,
  height: number,
  channels: 1 | 3,
): Float32Array {
  const out = new Float32Array(data.length);
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      for (let c = 0; c < channels; c++) {
        let sum = 0;
        let count = 0;
        for (let oy = -1; oy <= 1; oy++) {
          for (let ox = -1; ox <= 1; ox++) {
            const px = clampInt(x + ox, 0, width - 1);
            const py = clampInt(y + oy, 0, height - 1);
            sum += data[(py * width + px) * channels + c]!;
            count++;
          }
        }
        out[(y * width + x) * channels + c] = sum / count;
      }
    }
  }
  return out;
}

function clamp01(v: number): number {
  if (v <= 0) return 0;
  if (v >= 1) return 1;
  return v;
}

function clampInt(v: number, lo: number, hi: number): number {
  if (v < lo) return lo;
  if (v > hi) return hi;
  return v;
}
