// @a0n/.ignored_gnosis/distributed-inference-host/src/style-transfer-types.ts
import type { StyleTransferImagingWasmSurface } from './style-transfer-imaging-wasm';
import type { TopologyAlphaStyleBackend } from './style-transfer-topology-glue';

export type { StyleTransferImagingWasmSurface } from './style-transfer-imaging-wasm';

export type StyleTransferMethod = 'classical' | 'adain-adapter';

/** JSON provenance for mesh-gated style transfer (replay + telemetry). */
export interface StyleTransferMeshAdmissionChannel {
  readonly channelId: string;
  readonly meanSignal: number;
  readonly admissibleWeight: number;
}

/**
 * Replay-oriented mesh gate provenance. The six `channels[]` rows are
 * engineering telemetry (not Lean “six physics” coefficients):
 * `edge_coherence_wave`, `topology_sieve`, `depth_confidence`,
 * `photometric_veto`, `ensemble_disagreement`, `carrier_posture` — see
 * `STYLE_TRANSFER_MESH_SIX_CHANNEL_IDS` in `style-transfer-mesh-gate.ts`.
 */
export interface StyleTransferMeshProvenance {
  readonly schema_version: number;
  /** `wasm` when backend returned alpha; `stub` for host Sobel fingerprint. */
  readonly topology_source: 'wasm' | 'stub';
  readonly topology_provenance_fragment: Readonly<Record<string, unknown>>;
  readonly depth_provenance_fragment: Readonly<Record<string, unknown>>;
  readonly channels: ReadonlyArray<StyleTransferMeshAdmissionChannel>;
  readonly mean_effective_gate: number;
  readonly optional_depth_used: boolean;
}

/** Metrics used by fork/race/fold ranking beyond scalar styleDistance. */
export interface StyleTransferMeshFoldMetrics {
  readonly mean_topology_alpha: number;
  readonly mean_depth_confidence: number;
  readonly mean_effective_gate: number;
  readonly gate_violation_score: number;
}

export type StyleTransferImageInput =
  | string
  | Buffer
  | Uint8Array
  | {
      data: Float32Array | Uint8Array;
      width: number;
      height: number;
      channels: 1 | 3;
    };

export interface StyleTransferAnyImageOptions {
  method?: StyleTransferMethod;
  /** Blend between original content and stylized output. */
  strength?: number;
  /** Preserve content structure while moving tone/chroma to style. */
  structurePreservation?: number;
  /** Alpha-wave transposition controls. */
  alphaWave?:
    | boolean
    | {
        edgeThreshold?: number;
        edgeBoost?: number;
        frequency?: number;
        phase?: number;
      };
  /** Backward-compatible alias for alphaWave. */
  alphaOutline?: StyleTransferAnyImageOptions['alphaWave'];
  /** Optional sharpness lift after style application. */
  sharpen?: boolean;
  /**
   * Mesh admission gate: combines alpha-wave with topology (and optional depth)
   * so style strength is reduced where structure / depth is unreliable.
   */
  meshGate?: {
    enabled?: boolean;
    /** Per-pixel depth confidence in [0,1], length width*height, when available from WASM or native mesh. */
    depthConfidence?: Float32Array;
    /** Optional per-pixel veto in [0,1] (e.g. highlight clip guard); length width*height. */
    photometricVeto?: Float32Array;
    /**
     * When true and `imagingWasm.wasmMeshDepthConfidenceRgbF32` is present, fill
     * depth confidence from WASM (`build_mesh_depth_field` in the wasm crate).
     */
    autoWasmDepth?: boolean;
    /** Forwarded to wasm depth builder as monocular cue blend in `[0,1]`. Default 0. */
    wasmMonocularCueBlend?: number;
    /**
     * Host Gibson-style admission (0 = off). Blends with `photometricVeto` when set.
     */
    hostMonocularAdmissionWeight?: number;
    /**
     * When using `adain-adapter` with mesh gate: pull AdaIN output toward content
     * using scalar `mean_effective_gate` from the classical mesh path (per-pixel
     * gate on AdaIN is future work).
     */
    applyScalarMeshAdmissionToAdaIN?: boolean;
  };
}

export interface StyleTransferChannelStats {
  mean: number;
  stdDev: number;
}

export interface StyleTransferCandidateReport {
  label: string;
  methodUsed: StyleTransferMethod;
  rank: number;
  buleyWeight: number;
  rejectedByCeiling: boolean;
  styleDistance: number;
  /** Lexicographic tie-break / fold sort: lower is better when present. */
  compositeFoldScore?: number;
  meshFoldMetrics?: StyleTransferMeshFoldMetrics;
}

export interface StyleTransferAnyImageResult {
  stylized: Float32Array;
  width: number;
  height: number;
  channels: 1 | 3;
  methodUsed: StyleTransferMethod;
  styleDistance: number;
  strengthUsed: number;
  structurePreservationUsed: number;
  alphaWaveUsed: {
    enabled: boolean;
    edgeThreshold: number;
    edgeBoost: number;
    frequency: number;
    phase: number;
    meanAlpha: number;
  };
  contentStats: ReadonlyArray<StyleTransferChannelStats>;
  styleStats: ReadonlyArray<StyleTransferChannelStats>;
  outputStats: ReadonlyArray<StyleTransferChannelStats>;
  meshProvenance?: StyleTransferMeshProvenance;
  meshFoldMetrics?: StyleTransferMeshFoldMetrics;
}

export interface StyleTransferEnsembleConfig {
  label?: string;
  options?: StyleTransferAnyImageOptions;
}

export interface StyleTransferEnsembleArgs {
  content: StyleTransferImageInput;
  style: StyleTransferImageInput;
  configs: ReadonlyArray<StyleTransferEnsembleConfig>;
  concurrency?: number;
  ceiling?: number;
  /**
   * Optional topology WASM/native backend for all forks. If omitted, mesh gate
   * uses host stub when `meshGate.enabled` is set per-config.
   */
  topologyAlphaBackend?: TopologyAlphaStyleBackend;
  /** Penalty weights for fold ranking: score = styleDistance + wTopo*(1-mean_topology) + wDepth*(1-mean_depth). */
  meshFoldRankWeights?: { topology?: number; depth?: number };
  /**
   * Shared wasm imaging exports (`wasmTopologyMaskBuildRgbF32`, optional
   * `wasmMeshDepthConfidenceRgbF32`) for all forks when per-config backend is omitted.
   */
  imagingWasm?: StyleTransferImagingWasmSurface;
}

export interface StyleTransferEnsembleResult {
  stylized: Float32Array;
  width: number;
  height: number;
  channels: 1 | 3;
  candidates: ReadonlyArray<StyleTransferCandidateReport>;
  fold: {
    rounds: number;
    survivors: number;
    rejected: number;
    totalWeight: number;
  };
}
