// @a0n/.ignored_gnosis/distributed-inference-host/src/style-transfer-imaging-wasm.ts
/**
 * Optional wasm-bindgen surface for style-transfer imaging (topology mask,
 * mesh depth confidence). Pass the same namespace a Worker gets from
 * `distributed_inference.js` after `initSync` / default import.
 *
 * CI: integration tests can set `RUN_STYLE_TRANSFER_IMAGING_WASM_INTEGRATION=1`
 * and `DISTRIBUTED_INFERENCE_IMAGING_WASM_JS` to a resolved path for the glue
 * file (Node experimental loader or pre-bundled entry).
 */

import type { TopologyAlphaStyleBackend, TopologyAlphaStyleRequest } from './style-transfer-topology-glue';

/** wasm-bindgen result shape for topology mask. */
export interface WasmTopologyMaskResultLike {
  alpha(): Float32Array;
  provenance_json(): string;
  free?(): void;
}

export interface StyleTransferImagingWasmSurface {
  wasmTopologyMaskBuildRgbF32(
    rgb: Float32Array,
    width: number,
    height: number,
  ): WasmTopologyMaskResultLike;
  wasmMeshDepthConfidenceRgbF32?: (
    rgb: Float32Array,
    width: number,
    height: number,
    monocularCueBlend: number,
  ) => WasmMeshDepthStyleResultLike;
  wasmMeshResidualDeblurRgbF32?: (
    rgb: Float32Array,
    width: number,
    height: number,
    overpotential: number,
    admission_floor: number,
    saturation_limit: number,
    depth_confidence_blend: number,
  ) => WasmMeshResidualDeblurResultLike;
}

export interface WasmMeshDepthStyleResultLike {
  confidence(): Float32Array;
  provenance_json(): string;
  free?(): void;
}

export interface WasmMeshResidualDeblurResultLike {
  rgb: Float32Array;
  admission: Float32Array;
  certificate_json: string;
  free?(): void;
}

function safeJsonFragment(json: string): Record<string, unknown> {
  try {
    const o = JSON.parse(json) as unknown;
    return typeof o === 'object' && o !== null && !Array.isArray(o)
      ? (o as Record<string, unknown>)
      : { raw: json };
  } catch {
    return { parse_error: true, raw: json.slice(0, 512) };
  }
}

/** Builds {@link TopologyAlphaStyleBackend} from wasm exports (Worker / Node with wasm). */
export function topologyAlphaStyleBackendFromImagingWasm(
  wasm: StyleTransferImagingWasmSurface,
): TopologyAlphaStyleBackend {
  return {
    async buildTopologyAlphaForStyle(req: TopologyAlphaStyleRequest) {
      try {
        const res = wasm.wasmTopologyMaskBuildRgbF32(
          req.data instanceof Float32Array ? req.data : new Float32Array(req.data),
          req.width,
          req.height,
        );
        try {
          const alpha = new Float32Array(res.alpha());
          const provenanceFragment = safeJsonFragment(res.provenance_json());
          return {
            alpha,
            provenanceFragment: { ...provenanceFragment, wasm: 'wasmTopologyMaskBuildRgbF32' },
          };
        } finally {
          res.free?.();
        }
      } catch {
        return null;
      }
    },
  };
}

export interface WasmMeshDepthForStyleRequest {
  readonly data: Float32Array;
  readonly width: number;
  readonly height: number;
  readonly channels: 1 | 3;
  readonly monocularCueBlend: number;
}

/**
 * Handles the gnosis try Wasm Mesh Depth Confidence workflow.
 */
export async function tryWasmMeshDepthConfidence(
  wasm: StyleTransferImagingWasmSurface | undefined,
  req: WasmMeshDepthForStyleRequest,
): Promise<{ confidence: Float32Array; provenanceFragment: Record<string, unknown> } | null> {
  const fn = wasm?.wasmMeshDepthConfidenceRgbF32;
  if (fn === undefined) return null;
  try {
    const res = fn(req.data, req.width, req.height, req.monocularCueBlend);
    try {
      const confidence = new Float32Array(res.confidence());
      return {
        confidence,
        provenanceFragment: {
          ...safeJsonFragment(res.provenance_json()),
          wasm: 'wasmMeshDepthConfidenceRgbF32',
        },
      };
    } finally {
      res.free?.();
    }
  } catch {
    return null;
  }
}
