// @a0n/.ignored_gnosis/distributed-inference-host/src/style-transfer-adain-adapter.ts
import type { StyleTransferAnyImageResult } from './style-transfer-types';

export interface AdaINAdapterRequest {
  content: Float32Array;
  style: Float32Array;
  width: number;
  height: number;
  channels: 1 | 3;
  strength: number;
  structurePreservation: number;
}

export interface AdaINAdapterBackend {
  run(req: AdaINAdapterRequest): Promise<Float32Array>;
}

/**
 * Handles the gnosis run Ada INAdapter workflow.
 */
export async function runAdaINAdapter(
  req: AdaINAdapterRequest,
  backend: AdaINAdapterBackend | undefined,
  fallback: StyleTransferAnyImageResult,
): Promise<StyleTransferAnyImageResult> {
  if (backend === undefined) {
    return {
      ...fallback,
      methodUsed: 'adain-adapter',
    };
  }
  const output = await backend.run(req);
  const expectedLen = req.width * req.height * req.channels;
  if (output.length !== expectedLen) {
    throw new RangeError(
      `runAdaINAdapter: backend output length=${output.length} expected=${expectedLen}`,
    );
  }
  const clamped = new Float32Array(expectedLen);
  for (let i = 0; i < expectedLen; i++) {
    const v = output[i]!;
    clamped[i] = v <= 0 ? 0 : v >= 1 ? 1 : v;
  }
  return {
    ...fallback,
    stylized: clamped,
    methodUsed: 'adain-adapter',
  };
}
