// @a0n/.ignored_gnosis/distributed-inference-host/src/style-transfer-topology-glue.ts
/**
 * Topology alpha for mesh-gated style transfer.
 *
 * Full `wasmTopologyMaskBuildRgbF32` integration belongs in a caller-supplied
 * {@link TopologyAlphaStyleBackend}. When none is provided or the backend
 * returns null, the host uses {@link stubTopologyAlphaFromContentRgb} — a
 * deterministic Sobel-based fingerprint on luminance — so gates and
 * provenance remain replayable without claiming Betti/WASM parity.
 */

export const STYLE_TRANSFER_TOPOLOGY_GLUE_SCHEMA_VERSION = 1 as const;

export interface TopologyAlphaStyleRequest {
  readonly data: Float32Array;
  readonly width: number;
  readonly height: number;
  readonly channels: 1 | 3;
}

/** Optional WASM/native hook; return null to fall back to stub. */
export interface TopologyAlphaStyleBackend {
  buildTopologyAlphaForStyle(
    req: TopologyAlphaStyleRequest,
  ): Promise<TopologyAlphaStyleResult | null>;
}

export interface TopologyAlphaStyleResult {
  /** Per-pixel admission in [0, 1], length width * height. */
  readonly alpha: Float32Array;
  /** Small JSON-serializable fragment for provenance / replay. */
  readonly provenanceFragment: Record<string, unknown>;
}

/**
 * Host stub: normalized gradient magnitude on luminance (or single channel).
 * Not a Betti sieve; documented in mesh provenance as `source: "stub"`.
 */
export function stubTopologyAlphaFromContentRgb(req: TopologyAlphaStyleRequest): TopologyAlphaStyleResult {
  const { data, width, height, channels } = req;
  const n = width * height;
  const luma = channels === 1 ? data : rgbToLuminancePlane(data, width, height);
  const grad = sobelMagnitude01(luma, width, height);
  const alpha = new Float32Array(n);
  let sum = 0;
  for (let i = 0; i < n; i++) {
    const g = grad[i]!;
    alpha[i] = g;
    sum += g;
  }
  return {
    alpha,
    provenanceFragment: {
      kind: 'stub_topology_alpha',
      schema_version: STYLE_TRANSFER_TOPOLOGY_GLUE_SCHEMA_VERSION,
      method: 'sobel_luma_normalized',
      meanAlpha: sum / Math.max(1, n),
    },
  };
}

/**
 * Resolves the Topology Alpha For Style.
 */
export async function resolveTopologyAlphaForStyle(
  req: TopologyAlphaStyleRequest,
  backend: TopologyAlphaStyleBackend | undefined,
): Promise<{ result: TopologyAlphaStyleResult; source: 'wasm' | 'stub' }> {
  if (backend !== undefined) {
    const wasm = await backend.buildTopologyAlphaForStyle(req);
    if (wasm !== null) return { result: wasm, source: 'wasm' };
  }
  return { result: stubTopologyAlphaFromContentRgb(req), source: 'stub' };
}

function rgbToLuminancePlane(rgb: Float32Array, width: number, height: number): Float32Array {
  const n = width * height;
  const out = new Float32Array(n);
  for (let i = 0; i < n; i++) {
    const r = rgb[i * 3]!;
    const g = rgb[i * 3 + 1]!;
    const b = rgb[i * 3 + 2]!;
    out[i] = 0.2126 * r + 0.7152 * g + 0.0722 * b;
  }
  return out;
}

function sobelMagnitude01(luma: Float32Array, width: number, height: number): Float32Array {
  const gx = [-1, 0, 1, -2, 0, 2, -1, 0, 1];
  const gy = [1, 2, 1, 0, 0, 0, -1, -2, -1];
  const raw = new Float32Array(width * height);
  let maxMag = 1e-9;
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      let sx = 0;
      let sy = 0;
      let k = 0;
      for (let oy = -1; oy <= 1; oy++) {
        for (let ox = -1; ox <= 1; ox++, k++) {
          const px = clampInt(x + ox, 0, width - 1);
          const py = clampInt(y + oy, 0, height - 1);
          const v = luma[py * width + px]!;
          sx += gx[k]! * v;
          sy += gy[k]! * v;
        }
      }
      const m = Math.hypot(sx, sy);
      const p = y * width + x;
      raw[p] = m;
      if (m > maxMag) maxMag = m;
    }
  }
  const out = new Float32Array(width * height);
  const inv = 1 / maxMag;
  for (let i = 0; i < out.length; i++) {
    const v = raw[i]! * inv;
    out[i] = v <= 0 ? 0 : v >= 1 ? 1 : v;
  }
  return out;
}

function clampInt(v: number, lo: number, hi: number): number {
  if (v < lo) return lo;
  if (v > hi) return hi;
  return v;
}
