// @a0n/.ignored_gnosis/distributed-inference-host/src/style-transfer-any-image.ts
import { runAdaINAdapter, type AdaINAdapterBackend } from './style-transfer-adain-adapter';
import { runClassicalStyleTransfer } from './style-transfer-classical';
import {
  meshDepthProvenanceFromCallerBuffer,
  meshDepthProvenanceUnused,
  normalizePerPixelBuffer,
} from './style-transfer-mesh-depth-glue';
import { hostMonocularAdmissionFromRgb } from './style-transfer-monocular-host';
import {
  topologyAlphaStyleBackendFromImagingWasm,
  tryWasmMeshDepthConfidence,
  type StyleTransferImagingWasmSurface,
} from './style-transfer-imaging-wasm';
import { resolveTopologyAlphaForStyle, type TopologyAlphaStyleBackend } from './style-transfer-topology-glue';
import type {
  StyleTransferAnyImageOptions,
  StyleTransferAnyImageResult,
  StyleTransferImageInput,
} from './style-transfer-types';

interface DecodedImage {
  data: Float32Array;
  width: number;
  height: number;
  channels: 1 | 3;
}

/**
 * Handles the gnosis style Transfer Any Image workflow.
 */
export async function styleTransferAnyImage(
  contentInput: StyleTransferImageInput,
  styleInput: StyleTransferImageInput,
  options: StyleTransferAnyImageOptions = {},
  adainBackend?: AdaINAdapterBackend,
  topologyAlphaBackend?: TopologyAlphaStyleBackend,
  imagingWasm?: StyleTransferImagingWasmSurface,
): Promise<StyleTransferAnyImageResult> {
  const content = await decodeStyleImage(contentInput);
  const style = await decodeStyleImage(styleInput);
  if (
    content.width !== style.width ||
    content.height !== style.height ||
    content.channels !== style.channels
  ) {
    throw new RangeError(
      'styleTransferAnyImage: content/style dimensions and channels must match',
    );
  }

  const strength = clamp01(options.strength ?? 0.7);
  const structurePreservation = clamp01(options.structurePreservation ?? 0.55);
  const alphaWave = resolveAlphaWave(options.alphaWave ?? options.alphaOutline);
  const sharpen = options.sharpen ?? false;
  const method = options.method ?? 'classical';
  const meshGateOpts = options.meshGate;
  const meshGateEnabled = meshGateOpts !== undefined && meshGateOpts.enabled === true;
  const nPix = content.width * content.height;

  let meshGateArg: Parameters<typeof runClassicalStyleTransfer>[0]['meshGate'] = undefined;
  if (meshGateEnabled) {
    const topoBackend: TopologyAlphaStyleBackend | undefined =
      topologyAlphaBackend ??
      (imagingWasm?.wasmTopologyMaskBuildRgbF32 !== undefined
        ? topologyAlphaStyleBackendFromImagingWasm(imagingWasm)
        : undefined);
    const { result: topo, source: topologySource } = await resolveTopologyAlphaForStyle(
      {
        data: content.data,
        width: content.width,
        height: content.height,
        channels: content.channels,
      },
      topoBackend,
    );

    let depthBuf = normalizePerPixelBuffer(meshGateOpts.depthConfidence, nPix);
    let depthProv: Readonly<Record<string, unknown>> = meshDepthProvenanceUnused();
    if (depthBuf !== null) {
      let s = 0;
      for (let i = 0; i < depthBuf.length; i++) s += depthBuf[i]!;
      depthProv = meshDepthProvenanceFromCallerBuffer(s / depthBuf.length);
    } else if (meshGateOpts.autoWasmDepth === true && imagingWasm !== undefined) {
      const wasmDepth = await tryWasmMeshDepthConfidence(imagingWasm, {
        data: content.data,
        width: content.width,
        height: content.height,
        channels: content.channels,
        monocularCueBlend: clamp01(meshGateOpts.wasmMonocularCueBlend ?? 0),
      });
      if (wasmDepth !== null) {
        depthBuf = wasmDepth.confidence;
        depthProv = { ...wasmDepth.provenanceFragment, source: 'wasm_mesh_depth' };
      }
    }

    let vetoBuf = normalizePerPixelBuffer(meshGateOpts.photometricVeto, nPix);
    const monoW = meshGateOpts.hostMonocularAdmissionWeight ?? 0;
    if (monoW > 1e-6 && content.channels === 3) {
      const mono = hostMonocularAdmissionFromRgb(
        content.data,
        content.width,
        content.height,
        content.channels,
      );
      const combined = new Float32Array(nPix);
      for (let i = 0; i < nPix; i++) {
        const base = vetoBuf !== null ? vetoBuf[i]! : 1;
        const m = mono[i]!;
        const factor = (1 - monoW) + monoW * m;
        combined[i] = clamp01(base * factor);
      }
      vetoBuf = combined;
    }

    meshGateArg = {
      topologyAlpha: topo.alpha,
      depthConfidence: depthBuf,
      photometricVeto: vetoBuf,
      topologySource: topologySource === 'wasm' ? 'wasm' : 'stub',
      topologyProvenanceFragment: topo.provenanceFragment,
      depthProvenanceFragment: depthProv,
    };
  }

  const classical = runClassicalStyleTransfer({
    content: content.data,
    style: style.data,
    width: content.width,
    height: content.height,
    channels: content.channels,
    strength,
    structurePreservation,
    alphaWave,
    sharpen,
    meshGate: meshGateArg,
  });

  if (method === 'classical') return classical;

  const adapted = await runAdaINAdapter(
    {
      content: content.data,
      style: style.data,
      width: content.width,
      height: content.height,
      channels: content.channels,
      strength,
      structurePreservation,
    },
    adainBackend,
    classical,
  );

  if (
    meshGateEnabled &&
    meshGateOpts?.applyScalarMeshAdmissionToAdaIN === true &&
    classical.meshFoldMetrics !== undefined
  ) {
    const g = clamp01(classical.meshFoldMetrics.mean_effective_gate);
    const outPlane = new Float32Array(adapted.stylized.length);
    const cdata = content.data;
    for (let i = 0; i < outPlane.length; i++) {
      outPlane[i] = clamp01(cdata[i]! + g * (adapted.stylized[i]! - cdata[i]!));
    }
    return { ...adapted, stylized: outPlane };
  }

  return adapted;
}

/**
 * Handles the gnosis encode Styled As PNG workflow.
 */
export async function encodeStyledAsPNG(
  result: StyleTransferAnyImageResult,
): Promise<Buffer> {
  const sharp = (await import('sharp')).default;
  const n = result.width * result.height * result.channels;
  const bytes = new Uint8Array(n);
  for (let i = 0; i < n; i++) {
    const v = result.stylized[i]!;
    const c = v <= 0 ? 0 : v >= 1 ? 1 : v;
    bytes[i] = Math.round(c * 255);
  }
  return sharp(bytes, {
    raw: { width: result.width, height: result.height, channels: result.channels },
  })
    .png()
    .toBuffer();
}

async function decodeStyleImage(input: StyleTransferImageInput): Promise<DecodedImage> {
  if (
    typeof input === 'object' &&
    input !== null &&
    !Buffer.isBuffer(input) &&
    !(input instanceof Uint8Array) &&
    'data' in input
  ) {
    const channels = input.channels;
    if (channels !== 1 && channels !== 3) {
      throw new RangeError(`decodeStyleImage: channels must be 1 or 3, got ${channels}`);
    }
    const expected = input.width * input.height * channels;
    if (input.data.length !== expected) {
      throw new RangeError(
        `decodeStyleImage: data.length=${input.data.length} expected=${expected}`,
      );
    }
    if (input.data instanceof Float32Array) {
      return {
        data: new Float32Array(input.data),
        width: input.width,
        height: input.height,
        channels,
      };
    }
    const out = new Float32Array(expected);
    for (let i = 0; i < expected; i++) out[i] = input.data[i]! / 255;
    return { data: out, width: input.width, height: input.height, channels };
  }

  const sharp = (await import('sharp')).default;
  const source: string | Buffer =
    typeof input === 'string' ? input : Buffer.isBuffer(input) ? input : Buffer.from(input);
  const { data, info } = await sharp(source)
    .removeAlpha()
    .toColourspace('srgb')
    .raw()
    .toBuffer({ resolveWithObject: true });
  const channels = info.channels === 1 ? 1 : info.channels === 3 ? 3 : null;
  if (channels === null) {
    throw new RangeError(`decodeStyleImage: unsupported channels=${info.channels}`);
  }
  const expected = info.width * info.height * channels;
  if (data.length !== expected) {
    throw new RangeError(
      `decodeStyleImage: decoded bytes length=${data.length} expected=${expected}`,
    );
  }
  const out = new Float32Array(expected);
  for (let i = 0; i < expected; i++) out[i] = data[i]! / 255;
  return { data: out, width: info.width, height: info.height, channels };
}

function clamp01(v: number): number {
  if (v <= 0) return 0;
  if (v >= 1) return 1;
  return v;
}

function resolveAlphaWave(
  value: StyleTransferAnyImageOptions['alphaWave'],
): { enabled: boolean; edgeThreshold: number; edgeBoost: number; frequency: number; phase: number } {
  if (value === false) {
    return { enabled: false, edgeThreshold: 0.22, edgeBoost: 0.7, frequency: 2.2, phase: 0 };
  }
  if (value === undefined || value === true) {
    return { enabled: true, edgeThreshold: 0.22, edgeBoost: 0.7, frequency: 2.2, phase: 0 };
  }
  return {
    enabled: true,
    edgeThreshold: clamp01(value.edgeThreshold ?? 0.22),
    edgeBoost: clamp01(value.edgeBoost ?? 0.7),
    frequency: Math.max(0.1, value.frequency ?? 2.2),
    phase: value.phase ?? 0,
  };
}
