// @a0n/.ignored_gnosis/distributed-inference-host/src/thoth-oath-ledger-file-store.ts
import { mkdir, readFile, writeFile } from 'node:fs/promises';
import { dirname } from 'node:path';

import type { ThothOathLedgerStore, ThothOathRecord } from './thoth-oath-tools';

export class ThothOathFileLedgerStore implements ThothOathLedgerStore {
  constructor(private readonly filePath: string) {}

  get path(): string {
    return this.filePath;
  }

  async load(): Promise<readonly ThothOathRecord[]> {
    try {
      const raw = await readFile(this.filePath, 'utf8');
      const parsed = JSON.parse(raw) as unknown;
      if (!Array.isArray(parsed)) return [];
      return parsed.filter(isThothOathRecord);
    } catch (error) {
      if (isMissingFileError(error)) return [];
      throw error;
    }
  }

  async save(records: readonly ThothOathRecord[]): Promise<void> {
    await mkdir(dirname(this.filePath), { recursive: true });
    await writeFile(this.filePath, `${JSON.stringify(records, null, 2)}\n`, 'utf8');
  }
}

/**
 * Creates the Thoth Oath File Ledger Store.
 */
export function createThothOathFileLedgerStore(
  filePath: string,
): ThothOathFileLedgerStore {
  return new ThothOathFileLedgerStore(filePath);
}

function isMissingFileError(error: unknown): boolean {
  return (
    typeof error === 'object' &&
    error !== null &&
    'code' in error &&
    error.code === 'ENOENT'
  );
}

function isThothOathRecord(value: unknown): value is ThothOathRecord {
  if (typeof value !== 'object' || value === null) return false;
  const record = value as Partial<ThothOathRecord>;
  return (
    typeof record.id === 'string' &&
    typeof record.invariant === 'string' &&
    typeof record.scope === 'string' &&
    typeof record.consent === 'string' &&
    (record.status === 'proposed' ||
      record.status === 'active' ||
      record.status === 'cleared') &&
    record.source === 'thoth-self-management' &&
    typeof record.theoremRef === 'string' &&
    typeof record.createdAt === 'number' &&
    typeof record.durationMs === 'number' &&
    typeof record.expiresAt === 'number' &&
    typeof record.schedule === 'object' &&
    record.schedule !== null
  );
}
