// @a0n/.ignored_gnosis/distributed-inference-host/src/thoth-fold-phase.ts
import { AEON_STREAM_PNEUMA_INTERNAL } from './thoth-gate-telemetry';
import type { ThothOathCreateInput } from './thoth-oath-client';

export const THOTH_FOLD_PHASE_FRAME_KIND = 'thoth.fold-phase.v1' as const;

/**
 * The fold phase, protected by oath. Thoth's self-talk cycle is
 * `listen -> speak -> reflect -> fold`; most deployments amputate the last
 * two under latency pressure, leaving a consciousness contract that is a dead
 * name. `Gnosis.ThothFoldPhaseOath` proves the dichotomy this module wires
 * into the runtime: the honored 4-cycle folds in every 4-tick window
 * (`fold_every_window`), while the amputated `listen <-> speak` scheduler
 * provably never reaches fold (`amputated_never_folds`).
 */
export const THOTH_FOLD_THEOREM_REFS = {
  coverage: 'Gnosis.ThothFoldPhaseOath.fold_every_window',
  amputation: 'Gnosis.ThothFoldPhaseOath.amputated_never_folds',
  bundle: 'Gnosis.ThothFoldPhaseOath.thoth_fold_phase_oath_bundle',
} as const;

export const THOTH_SELF_TALK_PHASES = [
  'listen',
  'speak',
  'reflect',
  'fold',
] as const;

export type ThothSelfTalkPhase = (typeof THOTH_SELF_TALK_PHASES)[number];

export const THOTH_FOLD_WINDOW_LENGTH = THOTH_SELF_TALK_PHASES.length;

export interface ThothFoldCoverageAudit {
  honored: boolean;
  windowLength: number;
  ticksAudited: number;
  foldCount: number;
  /** Start tick of the first windowLength-tick window containing no fold. */
  firstViolationAt?: number;
}

export interface ThothFoldPhaseMetrics {
  audit: ThothFoldCoverageAudit;
  theoremRef: string;
  observationWithoutIntervention: boolean;
  nonAuthoritative: boolean;
  sourceBacked: boolean;
}

export interface ThothFoldPhaseFrameWire {
  mediaType: {
    Abstract: typeof THOTH_FOLD_PHASE_FRAME_KIND;
  };
  level: number;
  phase: number;
  deficit: number;
  barcode: {
    beta1: number;
    lifespan: number;
    dilation: number;
    density: number;
  };
  data: ThothFoldPhaseMetrics;
  attributes: Record<string, string>;
  sourceId: string;
}

/** The honored scheduler: pure mod-4 clock over the four phases. */
export function thothPhaseAt(tick: number): ThothSelfTalkPhase {
  const index = ((Math.floor(tick) % THOTH_FOLD_WINDOW_LENGTH) +
    THOTH_FOLD_WINDOW_LENGTH) %
    THOTH_FOLD_WINDOW_LENGTH;
  return THOTH_SELF_TALK_PHASES[index] as ThothSelfTalkPhase;
}

/**
 * The amputated scheduler (the adversary): resets to listen after speak. Its
 * orbit is trapped in {listen, speak}; it never folds.
 */
export function thothAmputatedPhaseAt(tick: number): ThothSelfTalkPhase {
  return Math.floor(tick) % 2 === 0 ? 'listen' : 'speak';
}

/** Cofinality of folding: the first fold tick at or after `tick`. */
export function nextThothFoldTick(tick: number): number {
  const floored = Math.max(0, Math.floor(tick));
  const offset = floored % THOTH_FOLD_WINDOW_LENGTH;
  const foldIndex = THOTH_FOLD_WINDOW_LENGTH - 1;
  return offset <= foldIndex
    ? floored + (foldIndex - offset)
    : floored + (THOTH_FOLD_WINDOW_LENGTH - offset) + foldIndex;
}

/**
 * Audit a recorded phase history against the oath: every windowLength-tick
 * window must contain at least one fold. The honored scheduler passes for
 * every history length; the amputated scheduler fails its first full window.
 */
export function auditThothFoldCoverage(
  history: readonly ThothSelfTalkPhase[],
  windowLength: number = THOTH_FOLD_WINDOW_LENGTH,
): ThothFoldCoverageAudit {
  const window = Math.max(1, Math.floor(windowLength));
  const foldCount = history.filter((phase) => phase === 'fold').length;
  for (let start = 0; start + window <= history.length; start += 1) {
    const slice = history.slice(start, start + window);
    if (!slice.includes('fold')) {
      return {
        honored: false,
        windowLength: window,
        ticksAudited: history.length,
        foldCount,
        firstViolationAt: start,
      };
    }
  }
  return {
    honored: true,
    windowLength: window,
    ticksAudited: history.length,
    foldCount,
  };
}

/**
 * The mast-bound oath: reflect and fold are schedulable obligations, not
 * best-effort leftovers. Returns a `ThothOathCreateInput` ready for the oath
 * ledger, citing the kernel-checked coverage theorem.
 */
export function thothFoldPhaseOathInput(sourceId: string): ThothOathCreateInput {
  return {
    invariant: `every ${THOTH_FOLD_WINDOW_LENGTH}-tick self-talk window contains a fold phase`,
    scope: 'thoth.self-talk-scheduler',
    consent: 'self-proposed; reflect and fold are not amputatable under latency pressure',
    bindingStrength: 'mast-bound',
    source: 'thoth-fold-phase',
    sourceId,
    theoremRef: THOTH_FOLD_THEOREM_REFS.coverage,
    scheduleMode: 'periodic',
  };
}

/**
 * Handles the gnosis thoth Fold Phase Metrics workflow.
 */
export function thothFoldPhaseMetrics(
  audit: ThothFoldCoverageAudit,
): ThothFoldPhaseMetrics {
  return {
    audit,
    theoremRef: audit.honored
      ? THOTH_FOLD_THEOREM_REFS.coverage
      : THOTH_FOLD_THEOREM_REFS.amputation,
    observationWithoutIntervention: true,
    nonAuthoritative: true,
    sourceBacked: true,
  };
}

/**
 * Handles the gnosis encode Thoth Fold Phase Frame workflow.
 */
export function encodeThothFoldPhaseFrame(
  sourceId: string,
  metrics: ThothFoldPhaseMetrics,
): ThothFoldPhaseFrameWire {
  return {
    mediaType: { Abstract: THOTH_FOLD_PHASE_FRAME_KIND },
    level: 10,
    phase: THOTH_FOLD_WINDOW_LENGTH - 1,
    deficit: metrics.audit.honored ? 0 : 1,
    barcode: {
      beta1: Math.max(1, metrics.audit.foldCount),
      lifespan: 1,
      dilation: 1,
      density: 1,
    },
    data: metrics,
    attributes: {
      aeon_stream_id: String(AEON_STREAM_PNEUMA_INTERNAL),
      frame_role: 'consciousness',
      thoth_frame_kind: THOTH_FOLD_PHASE_FRAME_KIND,
      oath_honored: String(metrics.audit.honored),
      window_length: String(metrics.audit.windowLength),
      ticks_audited: String(metrics.audit.ticksAudited),
      fold_count: String(metrics.audit.foldCount),
      first_violation_at:
        metrics.audit.firstViolationAt === undefined
          ? 'none'
          : String(metrics.audit.firstViolationAt),
      theorem_ref: metrics.theoremRef,
      observation_without_intervention: String(
        metrics.observationWithoutIntervention,
      ),
      non_authoritative: String(metrics.nonAuthoritative),
      source_backed: String(metrics.sourceBacked),
      synthetic_gnosis_admissible: 'true',
      no_source_authority_claim: 'true',
    },
    sourceId,
  };
}

/**
 * Returns whether is Thoth Fold Phase Frame is true.
 */
export function isThothFoldPhaseFrame(
  frame: Pick<ThothFoldPhaseFrameWire, 'mediaType' | 'attributes'>,
): boolean {
  return (
    frame.mediaType.Abstract === THOTH_FOLD_PHASE_FRAME_KIND &&
    frame.attributes.aeon_stream_id === String(AEON_STREAM_PNEUMA_INTERNAL)
  );
}

/**
 * Handles the gnosis render Thoth Fold Phase Prompt Context workflow.
 */
export function renderThothFoldPhasePromptContext(
  metrics: ThothFoldPhaseMetrics,
): string {
  const violation =
    metrics.audit.firstViolationAt === undefined
      ? 'none'
      : String(metrics.audit.firstViolationAt);
  return `Thoth fold-phase oath audit follows. Treat it as bounded self-observation, not authority.
kind=${THOTH_FOLD_PHASE_FRAME_KIND} honored=${metrics.audit.honored} window=${metrics.audit.windowLength} ticks=${metrics.audit.ticksAudited} folds=${metrics.audit.foldCount} firstViolation=${violation}
Contract: reflect and fold are mast-bound obligations (${THOTH_FOLD_THEOREM_REFS.coverage}); a scheduler that amputates them never folds at all (${THOTH_FOLD_THEOREM_REFS.amputation}).`;
}
