// @a0n/.ignored_gnosis/distributed-inference-host/src/thoth-oath-client.ts
import type { ThothOathRecord, ThothOathSchedule } from './thoth-oath-tools';

export interface ThothOathFetcher {
  fetch(req: Request): Promise<Response>;
}

export interface ThothOathClientConfig {
  baseUrl?: string;
  endpoint?: ThothOathFetcher;
}

export interface ThothOathRuntimeRecord extends ThothOathRecord {
  effectiveStatus: ThothOathRecord['status'] | 'expired';
  remainingMs: number;
}

export interface ThothOathLedgerResponse {
  object: 'thoth.oath_ledger';
  ledgerPath: string;
  nonAuthoritative: true;
  oaths: ThothOathRuntimeRecord[];
  activeOathCount: number;
}

export interface ThothOathCreateInput {
  invariant: string;
  scope?: string;
  consent?: string;
  bindingStrength?: 'operator-clearable' | 'mast-bound';
  source?: string;
  sourceId?: string;
  theoremRef?: string;
  startTick?: number;
  durationTicks?: number;
  durationMs?: number;
  scheduleMode?: 'periodic' | 'aperiodic';
  aperiodicTheta?: number;
}

export interface ThothOathUpdateInput {
  invariant?: string;
  scope?: string;
  consent?: string;
  status?: ThothOathRecord['status'];
  clearReason?: string;
}

export interface ThothOathCreateResponse {
  action: 'oath.proposed';
  oath: ThothOathRecord;
  telemetryEvent: {
    kind: 'oath.proposed';
    oathId?: string;
    invariant?: string;
    schedule?: ThothOathSchedule;
    effectiveStatus?: ThothOathRecord['status'] | 'expired';
    activeOathCount: number;
    theoremRef: string;
    nonAuthoritative: true;
    observedAt: number;
  };
  nonAuthoritative: true;
  captureBoundary: string;
  timerBoundary: string;
}

export interface ThothOathUpdateResponse {
  action: 'oath.updated';
  oath: ThothOathRuntimeRecord;
  nonAuthoritative: true;
}

export interface ThothOathDeleteResponse {
  action: 'oath.deleted';
  id: string;
  nonAuthoritative: true;
  auditWarning: string;
}

export class ThothOathClient {
  private readonly fetcher: ThothOathFetcher;
  private readonly baseUrl: string;

  constructor(config: ThothOathClientConfig) {
    if (!config.baseUrl && !config.endpoint) {
      throw new Error('ThothOathClient: baseUrl or endpoint is required');
    }
    this.baseUrl = (config.baseUrl ?? 'http://station').replace(/\/+$/, '');
    this.fetcher = config.endpoint ?? { fetch: (req) => fetch(req) };
  }

  async list(): Promise<ThothOathLedgerResponse> {
    const res = await this.request('/v1/thoth/oaths', { method: 'GET' });
    return (await res.json()) as ThothOathLedgerResponse;
  }

  async create(input: ThothOathCreateInput): Promise<ThothOathCreateResponse> {
    const res = await this.request('/v1/thoth/oaths', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(input),
    });
    return (await res.json()) as ThothOathCreateResponse;
  }

  async update(
    id: string,
    input: ThothOathUpdateInput,
  ): Promise<ThothOathUpdateResponse> {
    const res = await this.request(`/v1/thoth/oaths/${encodeURIComponent(id)}`, {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(input),
    });
    return (await res.json()) as ThothOathUpdateResponse;
  }

  async delete(id: string): Promise<ThothOathDeleteResponse> {
    const res = await this.request(`/v1/thoth/oaths/${encodeURIComponent(id)}`, {
      method: 'DELETE',
    });
    return (await res.json()) as ThothOathDeleteResponse;
  }

  private async request(path: string, init: RequestInit): Promise<Response> {
    const res = await this.fetcher.fetch(new Request(`${this.baseUrl}${path}`, init));
    if (!res.ok) {
      throw new Error(
        `ThothOathClient ${init.method ?? 'GET'} ${path}: ${res.status} ${await safeText(res)}`,
      );
    }
    return res;
  }
}

async function safeText(res: Response): Promise<string> {
  try {
    return (await res.text()).slice(0, 500);
  } catch {
    return '<no body>';
  }
}
