// @a0n/.ignored_gnosis/distributed-inference-host/src/thoth-place-tools.ts
// thoth-place-tools.ts — Thoth's SENSE-OF-PLACE tools: query the headless
// living-city sim (thoth-place-sim.ts) ON DEMAND instead of stuffing world
// state into the prompt. Same local-tool pattern as thoth-oath-tools: an
// in-process AgenticMcpClient composed with the network MCP by name routing.
//
// Five tools:
//   thoth/place.load       — load a city sim (by name, or the request observer)
//   thoth/place.sample     — who is where at an hour; busiest venues now
//   thoth/place.simulate   — worldsim results over a TIME WINDOW: an intra-day
//                            range ("8pm-9pm" → fromHour 20, toHour 21) and/or
//                            a span of days ("last week" → days 7)
//   thoth/place.recommend  — where to eat / get coffee / shop / visit, ranked
//                            by simulated foot traffic + proximity
//   thoth/place.query      — raw named-venue lookup by category/name
//
// Results are compact STRINGS (serializeToolResult passes strings through
// verbatim) — token-frugal for the 512-token thoth-content judgment budget.
// Honesty contract: simulated numbers are labeled as such; venue names come
// only from sovereign OSM data; absence is reported, never papered over.

import type {
  AgenticMcpClient,
  AgenticToolDefinition,
  ToolExecutionResult,
} from './agentic-chat';
import {
  sampleCity,
  simulateWindow,
  recommendPlaces,
  type ThothCitySim,
  type ThothPlaceSimHost,
} from './thoth-place-sim';
import { planTrip } from './thoth-place-route';

export const THOTH_PLACE_LOAD_TOOL = 'thoth/place.load' as const;
export const THOTH_PLACE_SAMPLE_TOOL = 'thoth/place.sample' as const;
export const THOTH_PLACE_SIMULATE_TOOL = 'thoth/place.simulate' as const;
export const THOTH_PLACE_RECOMMEND_TOOL = 'thoth/place.recommend' as const;
export const THOTH_PLACE_QUERY_TOOL = 'thoth/place.query' as const;
export const THOTH_PLACE_ROUTE_TOOL = 'thoth/place.route' as const;

const PROVENANCE =
  '(simulated foot-traffic from deterministic resident schedules; venues are source-backed sovereign OSM — non-authoritative; never invent places)';

export const THOTH_PLACE_TOOL_DEFINITIONS: readonly AgenticToolDefinition[] = [
  {
    type: 'function',
    function: {
      name: THOTH_PLACE_LOAD_TOOL,
      description:
        'Load the living-city simulation for a place: street graph, named venues from sovereign OSM, and a deterministic resident population. Defaults to the request observer location when no city is given.',
      parameters: {
        type: 'object',
        properties: {
          city: {
            type: 'string',
            description: 'Catalog city name (e.g. "Winters", "San Francisco").',
          },
          lat: { type: 'number', description: 'Observer latitude (WGS84).' },
          lon: { type: 'number', description: 'Observer longitude (WGS84).' },
        },
      },
    },
    source: 'provided',
    metadata: { thothSenseOfPlace: true },
  },
  {
    type: 'function',
    function: {
      name: THOTH_PLACE_SAMPLE_TOOL,
      description:
        'Sample the loaded city at one local-solar hour: how many residents are home / at work / shopping / at a park, and which venues are busiest right then.',
      parameters: {
        type: 'object',
        properties: {
          hour: {
            type: 'number',
            description: 'Local solar hour 0-24 (default: now).',
          },
          iso: {
            type: 'string',
            description: 'ISO timestamp alternative to hour.',
          },
        },
      },
    },
    source: 'provided',
    metadata: { thothSenseOfPlace: true },
  },
  {
    type: 'function',
    function: {
      name: THOTH_PLACE_SIMULATE_TOOL,
      description:
        'Generate living-city simulation results over a TIME WINDOW: an intra-day hour range (8pm-9pm → fromHour 20, toHour 21) and/or a span of days (last week → days 7). Returns exact per-venue visits/dwell and activity person-hours, with per-day texture (weekends differ).',
      parameters: {
        type: 'object',
        properties: {
          fromHour: {
            type: 'number',
            description: 'Window start, local solar hour 0-24 (default 0).',
          },
          toHour: {
            type: 'number',
            description:
              'Window end, local solar hour (default 24). May be ≤ fromHour for a past-midnight window like 23→1.',
          },
          days: {
            type: 'number',
            description:
              'How many days the window spans, ending today (1-14; 7 = last week). Default 1.',
          },
          limit: {
            type: 'number',
            description: 'Max venues listed (1-10, default 8).',
          },
        },
      },
    },
    activation: {
      patterns: [
        '\\b(simulate|model|over|during|between)\\b[\\s\\S]{0,60}\\b(week|day|hour|evening|morning|night|weekend)\\b',
        '\\blast\\s+(week|night|month|few days)\\b',
      ],
      keywords: ['simulate', 'last week', 'busiest', 'usually', 'typically'],
      threshold: 0.7,
    },
    source: 'provided',
    metadata: { thothSenseOfPlace: true },
  },
  {
    type: 'function',
    function: {
      name: THOTH_PLACE_RECOMMEND_TOOL,
      description:
        'Recommend places in the loaded city — where to eat, get coffee, shop, or what to visit — ranked by simulated resident foot traffic around the asked hour, then proximity to the observer.',
      parameters: {
        type: 'object',
        properties: {
          kind: {
            type: 'string',
            description: "One of 'eat' | 'coffee' | 'shop' | 'visit' (default 'eat').",
          },
          hour: {
            type: 'number',
            description: 'Local solar hour the recommendation is for (default: now).',
          },
          limit: { type: 'number', description: 'Max results (1-10, default 5).' },
        },
      },
    },
    activation: {
      patterns: [
        '\\b(where|what)\\b[\\s\\S]{0,50}\\b(eat|lunch|dinner|coffee|visit|see|shop|drink)\\b',
        '\\brecommend\\b[\\s\\S]{0,60}\\b(place|restaurant|cafe|bar|spot|park)\\b',
      ],
      keywords: [
        'where to eat',
        'restaurant',
        'hungry',
        'recommend',
        'visit',
        'coffee',
        'lunch',
        'dinner',
        'what to see',
      ],
      threshold: 0.7,
    },
    source: 'provided',
    metadata: { thothSenseOfPlace: true },
  },
  {
    type: 'function',
    function: {
      name: THOTH_PLACE_ROUTE_TOOL,
      description:
        "Plan a trip between places in the loaded city — distance, walking/driving time, and traffic-safety exposure per leg, routed over the real street graph by the aeon-transportation solver. Stops are venue names (or 'here'/'home' for the observer); multiple stops make a multi-point agenda; roundTrip returns to the first stop.",
      parameters: {
        type: 'object',
        properties: {
          stops: {
            type: 'array',
            description:
              "Ordered stop list: venue names and/or 'here'/'home'. 2+ entries.",
            items: { type: 'string' },
          },
          roundTrip: {
            type: 'boolean',
            description: 'Append a final leg back to the first stop.',
          },
        },
        required: ['stops'],
      },
    },
    activation: {
      patterns: [
        '\\bhow (far|long)\\b[\\s\\S]{0,60}\\b(to|from|between)\\b',
        '\\b(route|directions|get to|take me|walk to|drive to|there and back)\\b',
      ],
      keywords: [
        'how far',
        'how long',
        'route',
        'directions',
        'walk',
        'drive',
        'trip',
        'get there',
        'safely',
      ],
      threshold: 0.7,
    },
    source: 'provided',
    metadata: { thothSenseOfPlace: true },
  },
  {
    type: 'function',
    function: {
      name: THOTH_PLACE_QUERY_TOOL,
      description:
        'Look up named venues in the loaded city by category (food/shop/leisure/culture/health/...) and/or name substring. Raw listing, no ranking.',
      parameters: {
        type: 'object',
        properties: {
          category: { type: 'string', description: 'Venue category filter.' },
          name: { type: 'string', description: 'Case-insensitive name substring.' },
          limit: { type: 'number', description: 'Max results (1-20, default 10).' },
        },
      },
    },
    source: 'provided',
    metadata: { thothSenseOfPlace: true },
  },
];

export interface ThothPlaceToolOptions {
  host: ThothPlaceSimHost;
  /** Request observer — seeds the default city for sample/simulate/recommend/
   *  query when the model never called place.load explicitly. */
  observer?: { lat: number; lon: number; cityName?: string } | null;
  now?: () => number;
  randomId?: (prefix: string) => string;
}

const NO_CITY =
  'no city sense available: no city loaded and no observer to resolve one. Call thoth/place.load with a city name. Do not invent surroundings.';

function fmtHour(hour: number): string {
  const hh = Math.floor(hour);
  const mm = Math.round((hour - hh) * 60);
  return `${String(hh).padStart(2, '0')}:${String(mm).padStart(2, '0')}`;
}

function fmtDist(rec: { distanceMeters?: number; bearing?: string }): string {
  if (rec.distanceMeters === undefined) return '';
  return `, ~${rec.distanceMeters}m${rec.bearing ? ` ${rec.bearing}` : ''}`;
}

function clampNum(
  value: unknown,
  lo: number,
  hi: number,
  fallback: number
): number {
  const n = typeof value === 'number' && Number.isFinite(value) ? value : fallback;
  return Math.max(lo, Math.min(hi, n));
}

function optionalString(value: unknown): string | undefined {
  return typeof value === 'string' && value.trim().length > 0
    ? value.trim()
    : undefined;
}

function optionalNumber(value: unknown): number | undefined {
  return typeof value === 'number' && Number.isFinite(value)
    ? value
    : undefined;
}

export class ThothPlaceMcpClient implements AgenticMcpClient {
  readonly sessionId = 'thoth-place-local';
  private readonly host: ThothPlaceSimHost;
  private readonly observer:
    | { lat: number; lon: number; cityName?: string }
    | null;
  private readonly now: () => number;
  private readonly randomId: (prefix: string) => string;
  private sim: ThothCitySim | null = null;
  private simLoadAttempted = false;

  constructor(options: ThothPlaceToolOptions) {
    this.host = options.host;
    this.observer = options.observer ?? null;
    this.now = options.now ?? (() => Date.now());
    this.randomId =
      options.randomId ??
      ((prefix) => `${prefix}_${Math.random().toString(36).slice(2, 10)}`);
  }

  async listTools(): Promise<AgenticToolDefinition[]> {
    return [...THOTH_PLACE_TOOL_DEFINITIONS];
  }

  async preflight() {
    return {
      tools: await this.listTools(),
      sessionId: this.sessionId,
      durationMs: 0,
      cached: true,
    };
  }

  async callTool(
    name: string,
    args: Record<string, unknown>,
    options?: { toolCallId?: string }
  ): Promise<ToolExecutionResult> {
    const startedAt = this.now();
    const toolCallId = options?.toolCallId ?? this.randomId('call');
    const done = (result: string): ToolExecutionResult => ({
      toolCallId,
      toolName: name,
      args,
      result,
      success: true,
      durationMs: this.now() - startedAt,
    });
    try {
      if (name === THOTH_PLACE_LOAD_TOOL) {
        return done(await this.load(args));
      }
      if (name === THOTH_PLACE_SAMPLE_TOOL) {
        return done(await this.sample(args));
      }
      if (name === THOTH_PLACE_SIMULATE_TOOL) {
        return done(await this.simulate(args));
      }
      if (name === THOTH_PLACE_RECOMMEND_TOOL) {
        return done(await this.recommend(args));
      }
      if (name === THOTH_PLACE_QUERY_TOOL) {
        return done(await this.query(args));
      }
      if (name === THOTH_PLACE_ROUTE_TOOL) {
        return done(await this.route(args));
      }
    } catch (error) {
      return {
        toolCallId,
        toolName: name,
        args,
        result: null,
        success: false,
        durationMs: this.now() - startedAt,
        error: error instanceof Error ? error.message : String(error),
      };
    }
    return {
      toolCallId,
      toolName: name,
      args,
      result: null,
      success: false,
      durationMs: this.now() - startedAt,
      error: `Unknown Thoth place tool: ${name}`,
    };
  }

  /** Lazily resolve the active sim: explicit place.load wins; else the request
   *  observer; else honest absence. */
  private async ensureSim(): Promise<ThothCitySim | null> {
    if (this.sim) return this.sim;
    if (this.simLoadAttempted) return null;
    this.simLoadAttempted = true;
    if (!this.observer) return null;
    this.sim = await this.host.loadCitySim({
      ...(this.observer.cityName ? { city: this.observer.cityName } : {}),
      lat: this.observer.lat,
      lon: this.observer.lon,
    });
    return this.sim;
  }

  private async load(args: Record<string, unknown>): Promise<string> {
    const city = optionalString(args.city);
    const lat = optionalNumber(args.lat) ?? this.observer?.lat;
    const lon = optionalNumber(args.lon) ?? this.observer?.lon;
    if (!city && (lat === undefined || lon === undefined)) {
      return NO_CITY;
    }
    const sim = await this.host.loadCitySim({
      ...(city ? { city } : {}),
      ...(lat !== undefined ? { lat } : {}),
      ...(lon !== undefined ? { lon } : {}),
    });
    if (!sim) {
      return `no city sense available: could not load "${city ?? `${lat},${lon}`}" (unknown city or city graph unreachable). Do not invent surroundings.`;
    }
    this.sim = sim;
    this.simLoadAttempted = true;
    const employedPct = Math.round(
      (sim.residents.filter((r) => r.employed).length /
        Math.max(1, sim.residents.length)) *
        100
    );
    const hourNow = sampleCity(sim).hour;
    return [
      `loaded ${sim.cityName} (pop ${sim.population}, ${sim.source})`,
      `graph ${sim.nodes.size} plots | ${sim.venueCount} named venues on ${sim.placesByNode.size} plots | ${sim.residents.length} simulated residents (${employedPct}% employed)`,
      `local solar hour ${fmtHour(hourNow)} — use thoth/place.sample, thoth/place.simulate, thoth/place.recommend, thoth/place.query`,
    ].join('\n');
  }

  private async sample(args: Record<string, unknown>): Promise<string> {
    const sim = await this.ensureSim();
    if (!sim) return NO_CITY;
    const s = sampleCity(sim, {
      ...(optionalNumber(args.hour) !== undefined
        ? { hour: optionalNumber(args.hour)! }
        : {}),
      ...(optionalString(args.iso) ? { iso: optionalString(args.iso)! } : {}),
    });
    const activity = Object.entries(s.activity)
      .sort((a, b) => b[1] - a[1])
      .map(([k, v]) => `${k} ${v}`)
      .join(', ');
    const busiest =
      s.busiestVenues.length > 0
        ? `busiest venues now: ${s.busiestVenues
            .map((v) => `${v.name} (${v.category}) ${v.visits}`)
            .join(', ')}`
        : 'no named venues available — activity only. Do not invent venues.';
    return [
      `${sim.cityName} @ ${fmtHour(s.hour)} (${s.dayKey}) — ${s.residents} simulated residents: ${activity}`,
      busiest,
      PROVENANCE,
    ].join('\n');
  }

  private async simulate(args: Record<string, unknown>): Promise<string> {
    const sim = await this.ensureSim();
    if (!sim) return NO_CITY;
    const days = Math.round(clampNum(args.days, 1, 14, 1));
    const win = simulateWindow(sim, {
      ...(optionalNumber(args.fromHour) !== undefined
        ? { fromHour: optionalNumber(args.fromHour)! }
        : {}),
      ...(optionalNumber(args.toHour) !== undefined
        ? { toHour: optionalNumber(args.toHour)! }
        : {}),
      days,
      limit: Math.round(clampNum(args.limit, 1, 10, 8)),
    });
    const hours = Object.entries(win.activityPersonHours)
      .sort((a, b) => b[1] - a[1])
      .map(([k, v]) => `${k} ${v}h`)
      .join(', ');
    const lines = [
      `${sim.cityName} simulated ${fmtHour(win.fromHour)}–${fmtHour(win.toHour)} × ${win.days} day${win.days > 1 ? 's' : ''} (${win.residents} residents)`,
      `activity person-hours: ${hours}`,
    ];
    if (win.topVenues.length > 0) {
      lines.push(
        `venue visits ${win.venueVisitsTotal} total — top: ${win.topVenues
          .map(
            (v) =>
              `${v.name} (${v.category}) ${v.visits} visits` +
              (v.dwellHours ? ` · ${v.dwellHours}h dwell` : '')
          )
          .join(', ')}`
      );
    } else {
      lines.push(
        'no named venues available — activity only. Do not invent venues.'
      );
    }
    if (win.days > 1) {
      const busiest = [...win.perDay].sort(
        (a, b) => b.venueVisits - a.venueVisits
      )[0];
      const quietest = [...win.perDay].sort(
        (a, b) => a.venueVisits - b.venueVisits
      )[0];
      if (busiest && quietest) {
        lines.push(
          `per-day: busiest ${busiest.dayKey}${busiest.weekend ? ' (weekend)' : ''} ${busiest.venueVisits} visits · quietest ${quietest.dayKey}${quietest.weekend ? ' (weekend)' : ''} ${quietest.venueVisits}`
        );
      }
    }
    lines.push(PROVENANCE);
    return lines.join('\n');
  }

  private async recommend(args: Record<string, unknown>): Promise<string> {
    const sim = await this.ensureSim();
    if (!sim) return NO_CITY;
    const kindRaw = optionalString(args.kind)?.toLowerCase();
    const kind =
      kindRaw === 'coffee' || kindRaw === 'shop' || kindRaw === 'visit'
        ? kindRaw
        : 'eat';
    const recs = recommendPlaces(sim, {
      kind,
      ...(optionalNumber(args.hour) !== undefined
        ? { hour: optionalNumber(args.hour)! }
        : {}),
      limit: Math.round(clampNum(args.limit, 1, 10, 5)),
    });
    const hour =
      optionalNumber(args.hour) !== undefined
        ? fmtHour(((optionalNumber(args.hour)! % 24) + 24) % 24)
        : 'now';
    if (recs.length === 0) {
      return [
        `${kind} in ${sim.cityName} @ ${hour}: no named venues available in the loaded sim. Do not invent places.`,
        PROVENANCE,
      ].join('\n');
    }
    return [
      `${kind} in ${sim.cityName} @ ${hour}:`,
      ...recs.map(
        (r, i) =>
          `${i + 1}. ${r.name} (${r.category}${fmtDist(r)}) — ${r.why}`
      ),
      PROVENANCE,
    ].join('\n');
  }

  private async route(args: Record<string, unknown>): Promise<string> {
    const sim = await this.ensureSim();
    if (!sim) return NO_CITY;
    const stopsRaw = Array.isArray(args.stops) ? args.stops : [];
    const stops = stopsRaw
      .filter((s): s is string => typeof s === 'string' && s.trim().length > 0)
      .slice(0, 8);
    if (stops.length < 2) {
      return 'route needs at least 2 stops (venue names, or here/home for your location).';
    }
    const plan = await planTrip(sim, {
      stops,
      roundTrip: args.roundTrip === true,
    });
    const lines: string[] = [];
    if (plan.unresolved.length > 0) {
      lines.push(
        `unknown stops (not in the loaded city — do not invent them): ${plan.unresolved.join(', ')}`
      );
    }
    if (plan.solver === 'unavailable') {
      lines.push(
        'routing solver unavailable on this host — no distances. Do not estimate.'
      );
      return lines.join('\n');
    }
    if (plan.legs.length === 0) {
      lines.push('no routable legs (need 2+ resolvable stops).');
      return lines.join('\n');
    }
    lines.unshift(
      `trip in ${sim.cityName}${plan.roundTrip ? ' (round trip)' : ''} — ${plan.legs.length} leg${plan.legs.length > 1 ? 's' : ''}:`
    );
    for (const [i, leg] of plan.legs.entries()) {
      if (!leg.found) {
        lines.push(`${i + 1}. ${leg.from} → ${leg.to}: NO ROUTE on the street graph`);
        continue;
      }
      const safety =
        leg.safetyExposure !== null
          ? ` · safety exposure ${Math.round(leg.safetyExposure * 100) / 100}`
          : '';
      lines.push(
        `${i + 1}. ${leg.from} → ${leg.to}: ${leg.distanceKm}km` +
          (leg.walkMinutes !== null ? ` · walk ~${leg.walkMinutes}min` : '') +
          (leg.driveMinutes !== null ? ` · drive ~${leg.driveMinutes}min` : '') +
          safety
      );
    }
    const totals: string[] = [`total ${plan.totalKm}km`];
    if (plan.totalWalkMinutes !== null)
      totals.push(`walk ~${plan.totalWalkMinutes}min`);
    if (plan.totalDriveMinutes !== null)
      totals.push(`drive ~${plan.totalDriveMinutes}min`);
    lines.push(totals.join(' · '));
    lines.push(
      '(routed over the living-city street graph by aeon-transportation; times/safety are model estimates, not live traffic)'
    );
    return lines.join('\n');
  }

  private async query(args: Record<string, unknown>): Promise<string> {
    const sim = await this.ensureSim();
    if (!sim) return NO_CITY;
    const category = optionalString(args.category)?.toLowerCase();
    const nameSub = optionalString(args.name)?.toLowerCase();
    const limit = Math.round(clampNum(args.limit, 1, 20, 10));
    const seen = new Set<string>();
    const matches: string[] = [];
    let total = 0;
    for (const places of sim.placesByNode.values()) {
      for (const p of places) {
        if (category && p.category.toLowerCase() !== category) continue;
        if (nameSub && !p.name.toLowerCase().includes(nameSub)) continue;
        const key = p.name.toLowerCase();
        if (seen.has(key)) continue;
        seen.add(key);
        total += 1;
        if (matches.length < limit) {
          let dist = '';
          if (sim.observer) {
            const rec = {
              distanceMeters: Math.round(
                // haversine via recommendPlaces helpers would re-import; keep simple
                Math.hypot(
                  (p.lat - sim.observer.lat) * 111_320,
                  (p.lon - sim.observer.lon) *
                    111_320 *
                    Math.cos((sim.observer.lat * Math.PI) / 180)
                )
              ),
            };
            dist = fmtDist(rec);
          }
          matches.push(`${p.name} (${p.category}${dist})`);
        }
      }
    }
    if (total === 0) {
      return `no named venues match${category ? ` category=${category}` : ''}${nameSub ? ` name~${nameSub}` : ''} in ${sim.cityName}. Do not invent venues.`;
    }
    const tail = total > matches.length ? ` +${total - matches.length} more` : '';
    return `${sim.cityName} venues${category ? ` [${category}]` : ''}${nameSub ? ` ~"${nameSub}"` : ''}: ${matches.join('; ')}${tail}`;
  }
}

/**
 * Creates the Thoth Place Mcp Client.
 */
export function createThothPlaceMcpClient(
  options: ThothPlaceToolOptions
): ThothPlaceMcpClient {
  return new ThothPlaceMcpClient(options);
}
