// @a0n/.ignored_gnosis/distributed-inference-host/src/thoth-place-sim.ts
// thoth-place-sim.ts — the HEADLESS living-city simulation behind Thoth's
// sense-of-place tools (thoth-place-tools.ts) and the place-sense MCP.
//
// "Load a sim and sample it": resolve a catalog city → decode its street graph
// from city-graphs.knot (local file or a ~60KB ranged R2 fetch) → join real
// sovereign OSM venues onto plots (the proven Hotelling rule) → synthesize a
// deterministic resident population whose daily itineraries ARE the
// monster-studio living city's behavior (@a0n/aeon-3d/civic-schedule) → answer
// questions by enumerating itineraries:
//   sampleCity(hour)        — who is where right now; busiest venues
//   simulateWindow(window)  — EXACT visit/dwell aggregation over "8pm-9pm" or
//                             "last week" (stops are [start,end) hour segments,
//                             so window overlap enumerates every visit — no
//                             tick loop, no sampling error)
//   recommendPlaces(kind)   — eat/coffee/shop/visit ranked by simulated foot
//                             traffic, distance, name
//
// Day-to-day texture: schedules take a per-day variantSalt (ISO date) so a
// week isn't seven identical days, and weekends send most workers off-shift.
// Identity stays stable (cohort/employment derive from the unsalted id).
//
// HONESTY CONTRACT (the nearby-businesses rule): every stage degrades to null
// or honest absence — never a throw, never a fabricated venue. The population
// and its movements are SIMULATED and presented as such by the tool layer.

import {
  fetchSovereignPoiElements,
  haversineMeters,
  bearingDegrees,
  eightWindLabel,
  type SovereignPoiElement,
} from './thoth-nearby-businesses';
import {
  buildSchedule,
  stopAtHour,
  resolveDest,
  localSolarHourFromIso,
  fnv1a,
  type AgentSchedule,
  type ScheduleStop,
  type DestPools,
  type ScheduleGraph,
} from '@a0n/aeon-3d/civic-schedule';

// Heavy aeon-3d/aeon-solarium modules are loaded lazily inside loadCitySim —
// the solarium root barrel (via worldsim-earth-cities) is large, and a server
// that never uses place tools should not pay for it at boot.
type WorldCityLike = {
  name: string;
  lat: number;
  lon: number;
  population?: number;
};
type PlaceLike = {
  osmId: string;
  name: string;
  category: string;
  lat: number;
  lon: number;
  tags?: Readonly<Record<string, string>>;
};

export interface ThothPlaceGraphNode {
  id: string;
  position: { x: number; y: number; z: number };
  kind: string;
}

export interface ThothCitySim {
  cityName: string;
  lat: number;
  lon: number;
  population: number;
  source: 'catalog-knot' | 'synthetic-spiral';
  nodes: Map<string, ThothPlaceGraphNode>;
  /** Street edges (node-id pairs) — the routing surface for trip planning. */
  edges: Array<{ from: string; to: string; kind: string }>;
  /** Venues joined onto plots (Hotelling). Empty = honest "no named venues". */
  placesByNode: Map<string, PlaceLike[]>;
  venueCount: number;
  residents: SimResident[];
  pools: DestPools;
  observer?: { lat: number; lon: number };
  loadedAt: number;
}

export interface SimResident {
  id: string;
  employed: boolean;
  homeNodeId: string;
  workNodeId: string | null;
}

export interface VenueTally {
  nodeId: string;
  name: string;
  category: string;
  visits: number;
  /** Total dwell hours across visits (window mode). */
  dwellHours?: number;
  lat?: number;
  lon?: number;
}

export interface CitySample {
  hour: number;
  dayKey: string;
  residents: number;
  activity: Record<string, number>;
  busiestVenues: VenueTally[];
}

export interface WindowDayBreakdown {
  dayKey: string;
  weekend: boolean;
  venueVisits: number;
}

export interface WindowResult {
  fromHour: number;
  toHour: number;
  days: number;
  residents: number;
  /** Person-hours by activity over the whole window. */
  activityPersonHours: Record<string, number>;
  venueVisitsTotal: number;
  topVenues: VenueTally[];
  perDay: WindowDayBreakdown[];
}

export interface PlaceRecommendation {
  name: string;
  category: string;
  visits: number;
  distanceMeters?: number;
  bearing?: string;
  why: string;
}

export interface ThothPlaceSimHostOptions {
  fetchImpl?: typeof fetch;
  /** Local city-graphs.knot path (zero-network graph loading). */
  knotPath?: string;
  /** Edgework app base for the R2 knot proxy (world-knots bucket). */
  knotBaseUrl?: string;
  /** Sovereign POI base (/api/osm/poi). */
  poiBaseUrl?: string;
  maxCities?: number;
  maxAgents?: number;
  now?: () => number;
}

export interface ThothPlaceSimHost {
  loadCitySim(key: {
    city?: string;
    lat?: number;
    lon?: number;
  }): Promise<ThothCitySim | null>;
}

const ERRAND_CATEGORIES = new Set(['food', 'shop', 'leisure']);
const PARK_CATEGORIES = new Set(['leisure', 'culture']);
const VISIT_CATEGORIES = new Set(['leisure', 'culture', 'worship', 'civic']);

/** UTC day key used as the schedule variant salt (yyyy-mm-dd). */
export function dayKeyForDate(date: Date): string {
  return date.toISOString().slice(0, 10);
}

function isWeekend(dayKey: string): boolean {
  const d = new Date(`${dayKey}T12:00:00Z`).getUTCDay();
  return d === 0 || d === 6;
}

/** A resident's schedule for a given day: per-day salted timing/venues,
 *  stable identity, and weekends send ~75% of workers off-shift (leisure
 *  itinerary instead). */
export function scheduleForDay(
  resident: SimResident,
  dayKey: string
): AgentSchedule {
  let employedToday = resident.employed;
  if (employedToday && isWeekend(dayKey)) {
    employedToday =
      fnv1a(`${resident.id}|weekendoff|${dayKey}`) / 0x100000000 >= 0.75;
  }
  return buildSchedule(resident.id, employedToday, dayKey);
}

function overlapHours(
  stop: ScheduleStop,
  fromHour: number,
  toHour: number
): number {
  const a = Math.max(stop.startHour, fromHour);
  const b = Math.min(stop.endHour, toHour);
  return Math.max(0, b - a);
}

/** Venue attribution for a stop resolved to a node: a node can host several
 *  businesses (Hotelling cap 4) — pick deterministically by the stop's own
 *  pool hash so visits spread naturally and replay identically. */
function venueAtNode(
  sim: ThothCitySim,
  nodeId: string,
  stop: ScheduleStop
): PlaceLike | null {
  const places = sim.placesByNode.get(nodeId);
  if (!places || places.length === 0) return null;
  const h = typeof stop.destSel === 'object' ? stop.destSel.hash : 0;
  return places[h % places.length] ?? null;
}

const VISITABLE: ReadonlySet<string> = new Set([
  'breakfast',
  'dinner',
  'errand',
  'market',
  'charity',
  'festival',
  'lunch',
  'nightlife',
  'park',
  'school',
]);

interface VisitAccumulator {
  visits: number;
  dwellHours: number;
  name: string;
  category: string;
  nodeId: string;
  lat?: number;
  lon?: number;
}

/** Enumerate every venue visit committed by the population inside an
 *  intra-day window [fromHour, toHour) on a given day. EXACT: stops are hour
 *  segments, so overlap is a closed-form pass — no tick loop. */
function accumulateWindowDay(
  sim: ThothCitySim,
  dayKey: string,
  fromHour: number,
  toHour: number,
  activityPersonHours: Record<string, number>,
  venues: Map<string, VisitAccumulator>
): number {
  const graph: ScheduleGraph = { nodes: sim.nodes };
  let venueVisits = 0;
  for (const resident of sim.residents) {
    const schedule = scheduleForDay(resident, dayKey);
    for (const stop of schedule.stops) {
      const overlap = overlapHours(stop, fromHour, toHour);
      if (overlap <= 0) continue;
      activityPersonHours[stop.kind] =
        (activityPersonHours[stop.kind] ?? 0) + overlap;
      if (!VISITABLE.has(stop.kind)) continue;
      const nodeId = resolveDest(
        stop,
        resident.homeNodeId,
        resident.workNodeId,
        sim.pools,
        graph
      );
      if (!nodeId || nodeId === resident.homeNodeId) continue;
      const venue = venueAtNode(sim, nodeId, stop);
      if (!venue) continue;
      const key = `${nodeId}|${venue.osmId}`;
      let acc = venues.get(key);
      if (!acc) {
        acc = {
          visits: 0,
          dwellHours: 0,
          name: venue.name,
          category: venue.category,
          nodeId,
          lat: venue.lat,
          lon: venue.lon,
        };
        venues.set(key, acc);
      }
      acc.visits += 1;
      acc.dwellHours += overlap;
      venueVisits += 1;
    }
  }
  return venueVisits;
}

function sortTallies(venues: Map<string, VisitAccumulator>): VenueTally[] {
  return Array.from(venues.values())
    .sort(
      (a, b) =>
        b.visits - a.visits ||
        b.dwellHours - a.dwellHours ||
        (a.name < b.name ? -1 : a.name > b.name ? 1 : 0)
    )
    .map((v) => ({
      nodeId: v.nodeId,
      name: v.name,
      category: v.category,
      visits: v.visits,
      dwellHours: Math.round(v.dwellHours * 10) / 10,
      ...(v.lat !== undefined ? { lat: v.lat } : {}),
      ...(v.lon !== undefined ? { lon: v.lon } : {}),
    }));
}

/** Who is where at a single local-solar hour (today by default). */
export function sampleCity(
  sim: ThothCitySim,
  opts: { hour?: number; iso?: string; dayKey?: string } = {}
): CitySample {
  const iso = opts.iso ?? new Date().toISOString();
  const hour =
    typeof opts.hour === 'number' && Number.isFinite(opts.hour)
      ? ((opts.hour % 24) + 24) % 24
      : localSolarHourFromIso(iso, sim.lon);
  const dayKey = opts.dayKey ?? dayKeyForDate(new Date(iso));
  const graph: ScheduleGraph = { nodes: sim.nodes };
  const activity: Record<string, number> = {};
  const venues = new Map<string, VisitAccumulator>();
  for (const resident of sim.residents) {
    const schedule = scheduleForDay(resident, dayKey);
    const stop = stopAtHour(schedule, hour);
    activity[stop.kind] = (activity[stop.kind] ?? 0) + 1;
    if (!VISITABLE.has(stop.kind)) continue;
    const nodeId = resolveDest(
      stop,
      resident.homeNodeId,
      resident.workNodeId,
      sim.pools,
      graph
    );
    if (!nodeId || nodeId === resident.homeNodeId) continue;
    const venue = venueAtNode(sim, nodeId, stop);
    if (!venue) continue;
    const key = `${nodeId}|${venue.osmId}`;
    let acc = venues.get(key);
    if (!acc) {
      acc = {
        visits: 0,
        dwellHours: 0,
        name: venue.name,
        category: venue.category,
        nodeId,
        lat: venue.lat,
        lon: venue.lon,
      };
      venues.set(key, acc);
    }
    acc.visits += 1;
  }
  return {
    hour: Math.round(hour * 100) / 100,
    dayKey,
    residents: sim.residents.length,
    activity,
    busiestVenues: sortTallies(venues).slice(0, 8),
  };
}

/** Aggregate worldsim results over a TIME WINDOW — an intra-day range
 *  ("8pm-9pm" → fromHour 20, toHour 21) and/or a span of DAYS ("last week" →
 *  days 7). Multi-day windows aggregate each day's (possibly full-day) range
 *  with per-day schedule variants, so weeks have real weekday/weekend
 *  texture. */
export function simulateWindow(
  sim: ThothCitySim,
  opts: {
    fromHour?: number;
    toHour?: number;
    days?: number;
    /** Anchor date for the window's LAST day (default: today UTC). */
    endIso?: string;
    limit?: number;
  } = {}
): WindowResult {
  const days = Math.max(1, Math.min(14, Math.round(opts.days ?? 1)));
  let fromHour =
    typeof opts.fromHour === 'number' && Number.isFinite(opts.fromHour)
      ? ((opts.fromHour % 24) + 24) % 24
      : 0;
  let toHour =
    typeof opts.toHour === 'number' && Number.isFinite(opts.toHour)
      ? opts.toHour
      : 24;
  if (toHour <= fromHour && !(fromHour === 0 && toHour === 24)) {
    // "11pm-1am" style wrap: treat as [from,24) ∪ [0,to) within each day.
    toHour += 24;
  }
  toHour = Math.min(toHour, fromHour + 24);

  const end = opts.endIso ? new Date(opts.endIso) : new Date();
  const activityPersonHours: Record<string, number> = {};
  const venues = new Map<string, VisitAccumulator>();
  const perDay: WindowDayBreakdown[] = [];

  for (let d = days - 1; d >= 0; d--) {
    const date = new Date(end.getTime() - d * 86_400_000);
    const dayKey = dayKeyForDate(date);
    let dayVisits = 0;
    if (toHour > 24) {
      dayVisits += accumulateWindowDay(
        sim,
        dayKey,
        fromHour,
        24,
        activityPersonHours,
        venues
      );
      dayVisits += accumulateWindowDay(
        sim,
        dayKey,
        0,
        toHour - 24,
        activityPersonHours,
        venues
      );
    } else {
      dayVisits = accumulateWindowDay(
        sim,
        dayKey,
        fromHour,
        toHour,
        activityPersonHours,
        venues
      );
    }
    perDay.push({ dayKey, weekend: isWeekend(dayKey), venueVisits: dayVisits });
  }

  for (const k of Object.keys(activityPersonHours)) {
    activityPersonHours[k] = Math.round(activityPersonHours[k]! * 10) / 10;
  }
  const tallies = sortTallies(venues);
  return {
    fromHour,
    toHour: toHour > 24 ? toHour - 24 : toHour,
    days,
    residents: sim.residents.length,
    activityPersonHours,
    venueVisitsTotal: tallies.reduce((s, v) => s + v.visits, 0),
    topVenues: tallies.slice(0, Math.max(1, Math.min(10, opts.limit ?? 8))),
    perDay,
  };
}

/** Ranked place recommendations from simulated foot traffic. */
export function recommendPlaces(
  sim: ThothCitySim,
  opts: {
    kind?: 'eat' | 'coffee' | 'shop' | 'visit';
    hour?: number;
    iso?: string;
    limit?: number;
  } = {}
): PlaceRecommendation[] {
  const kind = opts.kind ?? 'eat';
  const limit = Math.max(1, Math.min(10, opts.limit ?? 5));
  // Foot traffic over a ±1.5h window around the asked hour gives steadier
  // rankings than a single instant (a venue's lunch rush is a window).
  const iso = opts.iso ?? new Date().toISOString();
  const hour =
    typeof opts.hour === 'number' && Number.isFinite(opts.hour)
      ? ((opts.hour % 24) + 24) % 24
      : localSolarHourFromIso(iso, sim.lon);
  const win = simulateWindow(sim, {
    fromHour: Math.max(0, hour - 1.5),
    toHour: Math.min(24, hour + 1.5),
    endIso: iso,
    limit: 10,
  });
  const byKey = new Map(win.topVenues.map((v) => [`${v.name}|${v.category}`, v]));

  const matches = (p: PlaceLike): boolean => {
    if (kind === 'eat') return p.category === 'food';
    if (kind === 'coffee') {
      return (
        p.category === 'food' &&
        (p.tags?.amenity === 'cafe' || /coffee|cafe|espresso/i.test(p.name))
      );
    }
    if (kind === 'shop') return p.category === 'shop';
    return VISIT_CATEGORIES.has(p.category);
  };

  const seen = new Set<string>();
  const candidates: PlaceRecommendation[] = [];
  for (const places of sim.placesByNode.values()) {
    for (const p of places) {
      if (!matches(p)) continue;
      const dedupe = p.name.toLowerCase();
      if (seen.has(dedupe)) continue;
      seen.add(dedupe);
      const tally = byKey.get(`${p.name}|${p.category}`);
      const visits = tally?.visits ?? 0;
      const rec: PlaceRecommendation = {
        name: p.name,
        category: p.category,
        visits,
        why:
          visits > 0
            ? `${visits}/${sim.residents.length} residents headed here around then`
            : 'quiet then',
      };
      if (
        sim.observer &&
        Number.isFinite(p.lat) &&
        Number.isFinite(p.lon)
      ) {
        rec.distanceMeters = Math.round(
          haversineMeters(sim.observer.lat, sim.observer.lon, p.lat, p.lon)
        );
        rec.bearing = eightWindLabel(
          bearingDegrees(sim.observer.lat, sim.observer.lon, p.lat, p.lon)
        );
      }
      candidates.push(rec);
    }
  }
  candidates.sort(
    (a, b) =>
      b.visits - a.visits ||
      (a.distanceMeters ?? Number.POSITIVE_INFINITY) -
        (b.distanceMeters ?? Number.POSITIVE_INFINITY) ||
      (a.name < b.name ? -1 : a.name > b.name ? 1 : 0)
  );
  return candidates.slice(0, limit);
}

// ─── City loading ────────────────────────────────────────────────────────────

/**
 * Creates the Thoth Place Sim Host.
 */
export function createThothPlaceSimHost(
  options: ThothPlaceSimHostOptions = {}
): ThothPlaceSimHost {
  const fetchImpl =
    options.fetchImpl ?? (typeof fetch === 'function' ? fetch : undefined);
  const maxCities = Math.max(1, options.maxCities ?? 4);
  const maxAgents = Math.max(20, options.maxAgents ?? 500);
  const now = options.now ?? (() => Date.now());

  const cache = new Map<string, ThothCitySim>();
  let cachedKnotHeader: unknown | null = null;
  let cachedKnotBytes: Uint8Array | null = null;

  const remember = (sim: ThothCitySim): ThothCitySim => {
    cache.delete(sim.cityName);
    cache.set(sim.cityName, sim);
    while (cache.size > maxCities) {
      const oldest = cache.keys().next().value as string | undefined;
      if (oldest === undefined) break;
      cache.delete(oldest);
    }
    return sim;
  };

  const loadCitySim = async (key: {
    city?: string;
    lat?: number;
    lon?: number;
  }): Promise<ThothCitySim | null> => {
    try {
      // Lazy heavy imports (solarium barrel via the catalog).
      const [earthCities, cityKnot, knotFetch, worldKnots, cityGraphMod, places] =
        await Promise.all([
          import('@a0n/aeon-3d/worldsim-earth-cities'),
          import('@a0n/aeon-3d/worldsim-city-knot'),
          import('@a0n/aeon-3d/worldsim-knot-fetch'),
          import('@a0n/aeon-solarium/world-knots'),
          import('@a0n/aeon-3d/worldsim-city-graph'),
          import('@a0n/aeon-3d/worldsim-places'),
        ]);

      // 1) Resolve the city.
      const catalog = earthCities.WORLD_CITY_CATALOG as readonly WorldCityLike[];
      let city: WorldCityLike | null = null;
      if (key.city && key.city.trim().length > 0) {
        const want = key.city.trim().toLowerCase();
        city =
          catalog.find((c) => c.name.toLowerCase() === want) ??
          catalog.find((c) => c.name.toLowerCase().startsWith(want)) ??
          null;
      }
      if (
        !city &&
        Number.isFinite(key.lat) &&
        Number.isFinite(key.lon)
      ) {
        let best: WorldCityLike | null = null;
        let bestM = Number.POSITIVE_INFINITY;
        for (const c of catalog) {
          const m = haversineMeters(key.lat!, key.lon!, c.lat, c.lon);
          if (m < bestM) {
            bestM = m;
            best = c;
          }
        }
        if (
          best &&
          bestM <=
            Math.max(
              earthCities.worldCityRadius(best as never) as number,
              30_000
            )
        ) {
          city = best;
        }
      }
      if (!city) return null;

      const cached = cache.get(city.name);
      if (cached) return remember(cached);

      // 2) Street graph: local knot file → ranged R2 fetch → synthetic spiral.
      type DecodedShape = {
        nodes: readonly { x: number; y: number; z: number; kind: string }[];
        edges: readonly { from: number; to: number; kind: string }[];
      };
      let decoded: DecodedShape | null = null;
      let source: ThothCitySim['source'] = 'catalog-knot';
      if (options.knotPath) {
        try {
          if (!cachedKnotBytes) {
            const fs = await import('node:fs/promises');
            cachedKnotBytes = new Uint8Array(
              await fs.readFile(options.knotPath)
            );
          }
          const header = cityKnot.readCityKnotHeader(cachedKnotBytes);
          decoded = cityKnot.decodeCityGraphRecord(
            cityKnot.sliceCityRecord(cachedKnotBytes, header, city.name)
          ) as DecodedShape;
        } catch {
          decoded = null;
        }
      }
      if (!decoded && fetchImpl) {
        try {
          const url = worldKnots.buildWorldKnotUrl('city-graphs.knot', {
            ...(options.knotBaseUrl
              ? { edgeworkAppBase: options.knotBaseUrl }
              : {}),
          });
          const got = await knotFetch.fetchKnotRecordRanged(url, city.name, {
            fetchFn: fetchImpl as never,
            ...(cachedKnotHeader ? { header: cachedKnotHeader as never } : {}),
          });
          if (got) {
            cachedKnotHeader = got.header;
            decoded = cityKnot.decodeCityGraphRecord(
              got.record
            ) as DecodedShape;
          }
        } catch {
          decoded = null;
        }
      }
      const nodes = new Map<string, ThothPlaceGraphNode>();
      const edges: Array<{ from: string; to: string; kind: string }> = [];
      if (decoded && decoded.nodes.length > 0) {
        decoded.nodes.forEach((n, i) =>
          nodes.set(String(i), {
            id: String(i),
            position: { x: n.x, y: n.y ?? 0, z: n.z },
            kind: n.kind ?? 'localParcel',
          })
        );
        for (const e of decoded.edges) {
          edges.push({
            from: String(e.from),
            to: String(e.to),
            kind: e.kind ?? 'localParcel',
          });
        }
      } else {
        // Synthetic golden-spiral fallback — parity with build-city-knot.ts /
        // tiger-city-fill.ts settlementCityParams (populationToCityRadius is
        // the same min(50km, max(100, sqrt(pop)*30)) formula as
        // worldCityRadius — see that module's parity note).
        source = 'synthetic-spiral';
        const pop = city.population ?? 7214;
        const radiusKm =
          (earthCities.worldCityRadius(city as never) as number) / 1000;
        const popMillions = pop / 1_000_000;
        const sprawl = Math.min(2, Math.max(1, Math.cbrt(popMillions * 10)));
        const depthCap = popMillions > 0.5 ? 4 : 3;
        const spiral = cityGraphMod.generateGoldenSpiralCity(
          { x: 0, y: 0, z: 0 },
          radiusKm,
          city.name,
          { sprawl, depthCap } as never
        );
        for (const [id, n] of spiral.nodes) {
          nodes.set(id, {
            id,
            position: { x: n.position.x, y: n.position.y, z: n.position.z },
            kind: n.kind,
          });
        }
        for (const e of spiral.edges) {
          edges.push({ from: e.from, to: e.to, kind: e.kind });
        }
      }
      if (nodes.size === 0) return null;

      // 3) Venues: sovereign POIs over the graph extent, Hotelling-joined.
      let minX = Infinity,
        maxX = -Infinity,
        minZ = Infinity,
        maxZ = -Infinity;
      for (const n of nodes.values()) {
        minX = Math.min(minX, n.position.x);
        maxX = Math.max(maxX, n.position.x);
        minZ = Math.min(minZ, n.position.z);
        maxZ = Math.max(maxZ, n.position.z);
      }
      let placesByNode = new Map<string, PlaceLike[]>();
      if (fetchImpl) {
        const bbox = places.bboxFromNodeExtent(city.lat, city.lon, {
          minX,
          maxX,
          minZ,
          maxZ,
        }) as { minLat: number; minLon: number; maxLat: number; maxLon: number };
        const elements = await fetchSovereignPoiElements({
          bbox: [bbox.minLat, bbox.minLon, bbox.maxLat, bbox.maxLon],
          limit: 600,
          ...(options.poiBaseUrl ? { baseUrl: options.poiBaseUrl } : {}),
          fetchImpl,
        });
        const placeList = poiElementsToPlaces(
          elements,
          places.categorizeTags as (
            tags: Readonly<Record<string, string>>
          ) => string
        );
        if (placeList.length > 0) {
          const joinNodes = Array.from(nodes.values()).map((n) => ({
            id: n.id,
            position: { x: n.position.x, z: n.position.z },
          }));
          placesByNode = places.assignPlacesByMeshHotelling(
            placeList as never,
            joinNodes,
            city.lat,
            city.lon
          ) as Map<string, PlaceLike[]>;
        }
      }
      let venueCount = 0;
      for (const list of placesByNode.values()) venueCount += list.length;

      // 4) Population: deterministic residents (identity from unsalted ids).
      const population = city.population ?? 7214;
      const residentCount = Math.max(
        20,
        Math.min(maxAgents, Math.round(population / 2000) || 20)
      );
      const businessNodeIds = Array.from(placesByNode.keys());
      const homeCandidates = Array.from(nodes.keys()).filter(
        (id) => !placesByNode.has(id)
      );
      const homePool = homeCandidates.length > 0 ? homeCandidates : Array.from(nodes.keys());
      const residents: SimResident[] = [];
      for (let i = 0; i < residentCount; i++) {
        const id = `${city.name}:resident:${i}`;
        const employed = fnv1a(`${id}|emp`) / 0x100000000 < 0.6;
        residents.push({
          id,
          employed,
          homeNodeId: homePool[fnv1a(`${id}|home`) % homePool.length]!,
          workNodeId:
            employed && businessNodeIds.length > 0
              ? businessNodeIds[fnv1a(`${id}|job`) % businessNodeIds.length]!
              : null,
        });
      }

      // 5) Pools.
      const errand: string[] = [];
      const community: string[] = [];
      const park: string[] = [];
      for (const [nodeId, list] of placesByNode) {
        if (list.some((p) => ERRAND_CATEGORIES.has(p.category)))
          errand.push(nodeId);
        if (list.some((p) => VISIT_CATEGORIES.has(p.category)))
          community.push(nodeId);
        if (list.some((p) => PARK_CATEGORIES.has(p.category)))
          park.push(nodeId);
      }

      const sim: ThothCitySim = {
        cityName: city.name,
        lat: city.lat,
        lon: city.lon,
        population,
        source,
        nodes,
        edges,
        placesByNode,
        venueCount,
        residents,
        pools: { errand, community: community.length > 0 ? community : errand, park },
        ...(Number.isFinite(key.lat) && Number.isFinite(key.lon)
          ? { observer: { lat: key.lat!, lon: key.lon! } }
          : {}),
        loadedAt: now(),
      };
      return remember(sim);
    } catch {
      return null; // honest absence — never throw into the tool loop
    }
  };

  return { loadCitySim };
}

function poiElementsToPlaces(
  elements: SovereignPoiElement[] | null,
  categorize: (tags: Readonly<Record<string, string>>) => string
): PlaceLike[] {
  if (!Array.isArray(elements)) return [];
  const out: PlaceLike[] = [];
  for (const el of elements) {
    if (!el || typeof el !== 'object') continue;
    const tags = (el.tags ?? {}) as Record<string, string>;
    const name = typeof tags.name === 'string' ? tags.name.trim() : '';
    if (!name) continue;
    const lat =
      typeof el.lat === 'number' && Number.isFinite(el.lat)
        ? el.lat
        : typeof el.center?.lat === 'number'
          ? el.center.lat
          : undefined;
    const lon =
      typeof el.lon === 'number' && Number.isFinite(el.lon)
        ? el.lon
        : typeof el.center?.lon === 'number'
          ? el.center.lon
          : undefined;
    if (lat === undefined || lon === undefined) continue;
    out.push({
      osmId: `${el.type ?? 'node'}/${el.id ?? out.length}`,
      name,
      category: categorize(tags),
      lat,
      lon,
      tags,
    });
  }
  return out;
}
