// @a0n/.ignored_gnosis/distributed-inference-host/src/thoth-place-route.ts
// thoth-place-route.ts — TRIP PLANNING over the loaded living-city sim, routed
// and scored by @a0n/aeon-transportation (the Rust/WASM Fungco/Dijkstra solver
// stack). Thoth knows about places (thoth-place-sim); this gets him — and the
// people he advises — THERE AND BACK: distance between stops, per-mode travel
// time (walking/driving/...), traffic-safety exposure and mortality risk per
// leg, and multi-point agendas (visit A, then B, then C, optionally round-trip
// home). One WASM call per leg: analyze_route_json = solve + civil ETA +
// per-mode economics + safety + mortality, all sourced from the same street
// graph the residents walk.
//
// HONESTY: routes exist only over the loaded graph; unknown stops are reported
// (never guessed); the solver's own `found:false` propagates as an honest
// "no route". Same degrade-to-absence contract as the rest of the place suite.

import { haversineMeters } from './thoth-nearby-businesses';
import type { ThothCitySim, ThothPlaceGraphNode } from './thoth-place-sim';

// ─── WASM solver (lazy init, web-target glue fed wasm bytes under node) ─────

interface TransportWasm {
  analyze_route_json(
    topologyJson: string,
    requestJson: string,
    optionsJson: string
  ): string;
}

let wasmPromise: Promise<TransportWasm | null> | null = null;

async function loadTransportWasm(): Promise<TransportWasm | null> {
  if (!wasmPromise) {
    wasmPromise = (async () => {
      try {
        const mod = (await import('@a0n/aeon-transportation')) as unknown as {
          default: (init?: unknown) => Promise<unknown>;
          initSync: (init: { module: unknown }) => unknown;
          analyze_route_json: TransportWasm['analyze_route_json'];
        };
        const { createRequire } = await import('node:module');
        const fs = await import('node:fs/promises');
        const require = createRequire(import.meta.url);
        const wasmPath = require.resolve(
          '@a0n/aeon-transportation/pkg/aeon_transportation_bg.wasm'
        );
        const bytes = await fs.readFile(wasmPath);
        mod.initSync({ module: bytes });
        return { analyze_route_json: mod.analyze_route_json };
      } catch {
        return null; // solver unavailable → tools degrade honestly
      }
    })();
  }
  return wasmPromise;
}

/** Test hook: reset the module-level WASM memo. */
export function __resetTransportWasmForTests(): void {
  wasmPromise = null;
}

// ─── Topology projection (sim street graph → solver TopologyGraph JSON) ─────

const topologyCache = new WeakMap<ThothCitySim, string>();

function edgeKm(a: ThothPlaceGraphNode, b: ThothPlaceGraphNode): number {
  // 1 world unit = 1 km in the city frame (latLonToLocalKm convention).
  return Math.hypot(
    a.position.x - b.position.x,
    a.position.z - b.position.z
  );
}

/** Serialize the sim's street graph as an aeon-transportation TopologyGraph.
 *  Edge weight = street length in km, bidirectional. Built once per sim. */
export function topologyJsonForSim(sim: ThothCitySim): string {
  const cached = topologyCache.get(sim);
  if (cached) return cached;
  const nodes = Array.from(sim.nodes.values()).map((n) => ({
    id: n.id,
    kind: n.kind || 'plot',
    label: n.id,
    centroid: null,
    bounds: null,
    attributes: {},
  }));
  const edges = sim.edges.map((e) => {
    const a = sim.nodes.get(e.from);
    const b = sim.nodes.get(e.to);
    return {
      from: e.from,
      to: e.to,
      // EdgeKind is kebab-case serde; 'traverses' is the road/street relation.
      kind: 'traverses',
      weight: a && b ? Math.max(0.01, edgeKm(a, b)) : 0.1,
      bidirectional: true,
      attributes: {},
    };
  });
  const json = JSON.stringify({
    id: `thoth-place:${sim.cityName}`,
    dialect: 'gg/topology',
    extension: 'living-city',
    sources: [],
    nodes,
    edges,
  });
  topologyCache.set(sim, json);
  return json;
}

// ─── Stop resolution ─────────────────────────────────────────────────────────

export interface ResolvedStop {
  query: string;
  nodeId: string;
  label: string;
}

const START_ALIASES = new Set(['home', 'here', 'start', 'observer', 'me']);

/** Resolve a stop query to a graph node: 'here'/'home' → the node nearest the
 *  observer (else the city hub '0'); otherwise a case-insensitive venue-name
 *  match (exact, then substring). Returns null when nothing matches — the
 *  caller reports it honestly. */
export function resolveStop(
  sim: ThothCitySim,
  query: string
): ResolvedStop | null {
  const q = query.trim();
  if (!q) return null;
  if (START_ALIASES.has(q.toLowerCase())) {
    let nodeId = sim.nodes.keys().next().value as string | undefined;
    if (sim.observer) {
      // Nearest plot to the observer in the lat/lon frame via venue anchors
      // when available, else the hub. Venue lat/lon is the only geo anchor the
      // sim carries; fall back to the hub node for venue-less sims.
      let bestM = Number.POSITIVE_INFINITY;
      for (const [nid, places] of sim.placesByNode) {
        const p = places[0];
        if (!p) continue;
        const m = haversineMeters(
          sim.observer.lat,
          sim.observer.lon,
          p.lat,
          p.lon
        );
        if (m < bestM) {
          bestM = m;
          nodeId = nid;
        }
      }
    }
    if (!nodeId) return null;
    return { query: q, nodeId, label: 'your location' };
  }
  const want = q.toLowerCase();
  let exact: ResolvedStop | null = null;
  let partial: ResolvedStop | null = null;
  for (const [nodeId, places] of sim.placesByNode) {
    for (const p of places) {
      const name = p.name.toLowerCase();
      if (name === want) {
        exact = { query: q, nodeId, label: p.name };
        break;
      }
      if (!partial && name.includes(want)) {
        partial = { query: q, nodeId, label: p.name };
      }
    }
    if (exact) break;
  }
  return exact ?? partial;
}

// ─── Trip planning ───────────────────────────────────────────────────────────

export interface TripLeg {
  from: string;
  to: string;
  found: boolean;
  distanceKm: number;
  walkMinutes: number | null;
  driveMinutes: number | null;
  /** 0..1-ish normalized exposure from the safety scorer (lower = safer). */
  safetyExposure: number | null;
  steps: number;
}

export interface TripPlan {
  stops: ResolvedStop[];
  unresolved: string[];
  legs: TripLeg[];
  roundTrip: boolean;
  totalKm: number;
  totalWalkMinutes: number | null;
  totalDriveMinutes: number | null;
  solver: 'aeon-transportation' | 'unavailable';
}

function minutesForMode(
  byMode: Array<{ mode?: string; economics?: Record<string, unknown> }>,
  mode: string
): number | null {
  const entry = byMode.find(
    (m) => String(m.mode ?? '').toLowerCase() === mode.toLowerCase()
  );
  const eco = entry?.economics as Record<string, unknown> | undefined;
  for (const key of [
    'travelTimeMinutes',
    'travel_time_minutes',
    'travelMinutes',
    'minutes',
    'travelTimeMin',
  ]) {
    const v = eco?.[key];
    if (typeof v === 'number' && Number.isFinite(v)) return v;
  }
  // Fall back to hours fields.
  for (const key of ['travelTimeHours', 'travel_time_hours', 'hours']) {
    const v = eco?.[key];
    if (typeof v === 'number' && Number.isFinite(v)) return v * 60;
  }
  return null;
}

function safetyOf(analysis: Record<string, unknown>): number | null {
  const safety = analysis.safety as Record<string, unknown> | undefined;
  for (const key of [
    'exposure',
    'normalizedExposure',
    'normalized_exposure',
    'exposureScore',
    'score',
  ]) {
    const v = safety?.[key];
    if (typeof v === 'number' && Number.isFinite(v)) return v;
  }
  return null;
}

const WALK_KMH = 4.8;
const DRIVE_KMH = 32; // small-town street driving

/** Plan a multi-stop trip over the sim's street graph. Each consecutive pair
 *  becomes a solver leg; roundTrip appends a final leg back to the first
 *  stop. When the WASM solver is unavailable the plan still returns honest
 *  distances (graph-free haversine is NOT used — legs are marked not-found). */
export async function planTrip(
  sim: ThothCitySim,
  opts: { stops: string[]; roundTrip?: boolean }
): Promise<TripPlan> {
  const resolved: ResolvedStop[] = [];
  const unresolved: string[] = [];
  for (const q of opts.stops) {
    const r = resolveStop(sim, q);
    if (r) resolved.push(r);
    else unresolved.push(q);
  }
  const route = [...resolved];
  const roundTrip = Boolean(opts.roundTrip);
  if (roundTrip && route.length >= 2) route.push(route[0]!);

  const wasm = await loadTransportWasm();
  const legs: TripLeg[] = [];
  if (wasm && route.length >= 2) {
    const topology = topologyJsonForSim(sim);
    for (let i = 0; i < route.length - 1; i++) {
      const from = route[i]!;
      const to = route[i + 1]!;
      let leg: TripLeg = {
        from: from.label,
        to: to.label,
        found: false,
        distanceKm: 0,
        walkMinutes: null,
        driveMinutes: null,
        safetyExposure: null,
        steps: 0,
      };
      try {
        const raw = wasm.analyze_route_json(
          topology,
          JSON.stringify({ from: from.nodeId, to: to.nodeId }),
          '{}'
        );
        const analysis = JSON.parse(raw) as Record<string, unknown>;
        const found = analysis.found === true;
        const distanceKm =
          typeof analysis.cost === 'number' && Number.isFinite(analysis.cost)
            ? analysis.cost
            : 0;
        const byMode = Array.isArray(analysis.by_mode)
          ? (analysis.by_mode as Array<{
              mode?: string;
              economics?: Record<string, unknown>;
            }>)
          : [];
        const walk =
          minutesForMode(byMode, 'walking') ??
          (found ? (distanceKm / WALK_KMH) * 60 : null);
        const drive =
          minutesForMode(byMode, 'driving') ??
          (found ? (distanceKm / DRIVE_KMH) * 60 : null);
        leg = {
          from: from.label,
          to: to.label,
          found,
          distanceKm: Math.round(distanceKm * 100) / 100,
          walkMinutes: walk !== null ? Math.round(walk * 10) / 10 : null,
          driveMinutes: drive !== null ? Math.round(drive * 10) / 10 : null,
          safetyExposure: safetyOf(analysis),
          steps: Array.isArray(analysis.steps)
            ? (analysis.steps as unknown[]).length
            : 0,
        };
      } catch {
        // leg stays not-found — honest absence
      }
      legs.push(leg);
    }
  }

  const foundLegs = legs.filter((l) => l.found);
  const totalKm =
    Math.round(foundLegs.reduce((s, l) => s + l.distanceKm, 0) * 100) / 100;
  const sumOrNull = (pick: (l: TripLeg) => number | null): number | null => {
    if (foundLegs.length === 0) return null;
    let s = 0;
    for (const l of foundLegs) {
      const v = pick(l);
      if (v === null) return null;
      s += v;
    }
    return Math.round(s * 10) / 10;
  };
  return {
    stops: resolved,
    unresolved,
    legs,
    roundTrip,
    totalKm,
    totalWalkMinutes: sumOrNull((l) => l.walkMinutes),
    totalDriveMinutes: sumOrNull((l) => l.driveMinutes),
    solver: wasm ? 'aeon-transportation' : 'unavailable',
  };
}
