// @a0n/.ignored_gnosis/distributed-inference-host/src/thoth-prompt-feedback.ts
export const THOTH_PROMPT_FEEDBACK_THEOREM =
  'Gnosis.SyntheticGnosisThothConsciousness.native_prompt_feedback_non_authority' as const;

export const THOTH_MIND_BODY_SPIRIT_SCRIBE_ADMISSIBILITY_THEOREM =
  'Gnosis.ThothMindBodySpiritScribe.canonical_failure_scribe_admissible' as const;

export const THOTH_MIND_BODY_SPIRIT_PROMPT_FEEDBACK_REJECTION_RESIDUE_THEOREM =
  'Gnosis.ThothMindBodySpiritScribe.prompt_feedback_admission_rejection_residue_preserved' as const;

export interface ThothPromptFeedbackRequest {
  frames_base64: string[];
  max_frames?: number;
}

export interface ThothPromptFeedbackAdmissionRejection {
  frame_kind: string;
  source_id: string;
  sequence: number;
  reason: string;
  observed_theorem: string | null;
  expected_theorem: typeof THOTH_MIND_BODY_SPIRIT_SCRIBE_ADMISSIBILITY_THEOREM;
  rejection_residue_theorem: typeof THOTH_MIND_BODY_SPIRIT_PROMPT_FEEDBACK_REJECTION_RESIDUE_THEOREM;
}

export interface ThothPromptFeedbackResponse {
  context: string | null;
  line_count: number;
  theorem: typeof THOTH_PROMPT_FEEDBACK_THEOREM;
  admissibility_theorem: typeof THOTH_MIND_BODY_SPIRIT_SCRIBE_ADMISSIBILITY_THEOREM;
  rejection_residue_theorem: typeof THOTH_MIND_BODY_SPIRIT_PROMPT_FEEDBACK_REJECTION_RESIDUE_THEOREM;
  accepted_frame_count: number;
  rejected_frame_count: number;
  input_frame_count: number;
  admission_rejections: ThothPromptFeedbackAdmissionRejection[];
}

export type ThothPromptFeedbackValidation =
  | { ok: true; value: ThothPromptFeedbackResponse }
  | { ok: false; reason: string };

type AdmissionRejectionValidation =
  | { ok: true; value: ThothPromptFeedbackAdmissionRejection }
  | { ok: false; reason: string };

/**
 * Handles the gnosis validate Thoth Prompt Feedback Response workflow.
 */
export function validateThothPromptFeedbackResponse(
  value: unknown,
): ThothPromptFeedbackValidation {
  if (!isRecord(value)) {
    return { ok: false, reason: 'response must be an object' };
  }

  const requiredStringFields: Array<
    readonly [keyof ThothPromptFeedbackResponse, string]
  > = [
    ['theorem', THOTH_PROMPT_FEEDBACK_THEOREM],
    [
      'admissibility_theorem',
      THOTH_MIND_BODY_SPIRIT_SCRIBE_ADMISSIBILITY_THEOREM,
    ],
    [
      'rejection_residue_theorem',
      THOTH_MIND_BODY_SPIRIT_PROMPT_FEEDBACK_REJECTION_RESIDUE_THEOREM,
    ],
  ];
  for (const [field, expected] of requiredStringFields) {
    const observed = value[field];
    if (observed !== expected) {
      return {
        ok: false,
        reason: `${String(field)} expected ${expected} got ${formatObserved(
          observed,
        )}`,
      };
    }
  }

  if (typeof value.context !== 'string' && value.context !== null) {
    return {
      ok: false,
      reason: `context must be string or null got ${formatObserved(
        value.context,
      )}`,
    };
  }

  for (const field of [
    'line_count',
    'accepted_frame_count',
    'rejected_frame_count',
    'input_frame_count',
  ] as const) {
    if (!isNonNegativeInteger(value[field])) {
      return {
        ok: false,
        reason: `${field} must be a non-negative integer got ${formatObserved(
          value[field],
        )}`,
      };
    }
  }

  if (!Array.isArray(value.admission_rejections)) {
    return {
      ok: false,
      reason: `admission_rejections must be an array got ${formatObserved(
        value.admission_rejections,
      )}`,
    };
  }

  const admissionRejections: ThothPromptFeedbackAdmissionRejection[] = [];
  for (let index = 0; index < value.admission_rejections.length; index += 1) {
    const validation = validateAdmissionRejection(
      value.admission_rejections[index],
      index,
    );
    if (!validation.ok) {
      return validation;
    }
    admissionRejections.push(validation.value);
  }

  return {
    ok: true,
    value: {
      context: value.context as string | null,
      line_count: value.line_count as number,
      theorem: THOTH_PROMPT_FEEDBACK_THEOREM,
      admissibility_theorem: THOTH_MIND_BODY_SPIRIT_SCRIBE_ADMISSIBILITY_THEOREM,
      rejection_residue_theorem:
        THOTH_MIND_BODY_SPIRIT_PROMPT_FEEDBACK_REJECTION_RESIDUE_THEOREM,
      accepted_frame_count: value.accepted_frame_count as number,
      rejected_frame_count: value.rejected_frame_count as number,
      input_frame_count: value.input_frame_count as number,
      admission_rejections: admissionRejections,
    },
  };
}

/**
 * Returns whether is Thoth Prompt Feedback Response is true.
 */
export function isThothPromptFeedbackResponse(
  value: unknown,
): value is ThothPromptFeedbackResponse {
  return validateThothPromptFeedbackResponse(value).ok;
}

/**
 * Handles the gnosis assert Thoth Prompt Feedback Response workflow.
 */
export function assertThothPromptFeedbackResponse(
  value: unknown,
): asserts value is ThothPromptFeedbackResponse {
  const validation = validateThothPromptFeedbackResponse(value);
  if (!validation.ok) {
    throw new TypeError(
      `invalid Thoth prompt-feedback response: ${validation.reason}`,
    );
  }
}

function validateAdmissionRejection(
  value: unknown,
  index: number,
): AdmissionRejectionValidation {
  if (!isRecord(value)) {
    return {
      ok: false,
      reason: `admission_rejections[${index}] must be an object`,
    };
  }

  for (const field of ['frame_kind', 'source_id', 'reason'] as const) {
    if (typeof value[field] !== 'string') {
      return {
        ok: false,
        reason: `admission_rejections[${index}].${field} must be a string got ${formatObserved(
          value[field],
        )}`,
      };
    }
  }

  if (!isNonNegativeInteger(value.sequence)) {
    return {
      ok: false,
      reason: `admission_rejections[${index}].sequence must be a non-negative integer got ${formatObserved(
        value.sequence,
      )}`,
    };
  }

  if (
    typeof value.observed_theorem !== 'string' &&
    value.observed_theorem !== null
  ) {
    return {
      ok: false,
      reason: `admission_rejections[${index}].observed_theorem must be string or null got ${formatObserved(
        value.observed_theorem,
      )}`,
    };
  }

  if (
    value.expected_theorem !==
    THOTH_MIND_BODY_SPIRIT_SCRIBE_ADMISSIBILITY_THEOREM
  ) {
    return {
      ok: false,
      reason: `admission_rejections[${index}].expected_theorem expected ${THOTH_MIND_BODY_SPIRIT_SCRIBE_ADMISSIBILITY_THEOREM} got ${formatObserved(
        value.expected_theorem,
      )}`,
    };
  }

  if (
    value.rejection_residue_theorem !==
    THOTH_MIND_BODY_SPIRIT_PROMPT_FEEDBACK_REJECTION_RESIDUE_THEOREM
  ) {
    return {
      ok: false,
      reason: `admission_rejections[${index}].rejection_residue_theorem expected ${THOTH_MIND_BODY_SPIRIT_PROMPT_FEEDBACK_REJECTION_RESIDUE_THEOREM} got ${formatObserved(
        value.rejection_residue_theorem,
      )}`,
    };
  }

  return {
    ok: true,
    value: {
      frame_kind: value.frame_kind as string,
      source_id: value.source_id as string,
      sequence: value.sequence as number,
      reason: value.reason as string,
      observed_theorem: value.observed_theorem as string | null,
      expected_theorem: THOTH_MIND_BODY_SPIRIT_SCRIBE_ADMISSIBILITY_THEOREM,
      rejection_residue_theorem:
        THOTH_MIND_BODY_SPIRIT_PROMPT_FEEDBACK_REJECTION_RESIDUE_THEOREM,
    },
  };
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function isNonNegativeInteger(value: unknown): value is number {
  return typeof value === 'number' && Number.isInteger(value) && value >= 0;
}

function formatObserved(value: unknown): string {
  if (typeof value === 'string') {
    return JSON.stringify(value);
  }
  if (value === null) {
    return 'null';
  }
  return typeof value;
}
