// @a0n/.ignored_gnosis/distributed-inference-host/src/transcribe-audio.ts
/**
 * OpenAI-compatible /v1/audio/transcriptions handler, powered by the sovereign
 * in-process Whisper STT (`@a0n/voice-stt`, node target). Completes the thoth
 * voice line: chat + speech + transcriptions, all sovereign, no third party.
 *
 * Wire shape mirrors OpenAI's Whisper API: POST multipart/form-data carrying a
 * `file` part (the audio bytes), plus `model`, optional `language` and
 * `response_format` (json | text). Audio is decoded to 16 kHz mono f32 via
 * ffmpeg, then handed to a lazily-initialized WhisperEngine singleton.
 */

import { spawn } from 'node:child_process';
import {
  existsSync,
  mkdtempSync,
  readFileSync,
  rmSync,
  writeFileSync,
} from 'node:fs';
import { tmpdir } from 'node:os';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
// Sovereign in-browser/in-process Whisper STT compiled to WASM (node target).
// Lazy `import('@a0n/voice-stt')` -> `new stt.WhisperEngine()`; load_model(Uint8Array);
// transcribe(Float32Array, 16000) -> { text, confidence, language, segments }.
// LAZY on purpose: the package's node export is the wasm-pack `pkg-node/`
// artifact, which is not present in every image build (it crashed the whole
// openai-server at module load on the GPU moonshine image — chat died for
// STT's sake). With the dynamic import, a missing engine 500s only the
// transcription route; every other route serves.
type SttModule = typeof import('@a0n/voice-stt');
type WhisperEngine = InstanceType<SttModule['WhisperEngine']>;

export interface TranscriptionHttpResponse {
  status: number;
  headers: Record<string, string>;
  body: Uint8Array;
}

export type TranscriptionResponseFormat = 'json' | 'text';

export interface TranscriptionRequestInput {
  /** Raw HTTP request body bytes (multipart/form-data). */
  body: Uint8Array;
  /** HTTP request headers (need Content-Type for the multipart boundary). */
  headers: Record<string, string | string[] | undefined>;
  /** Override the WhisperEngine weights path (else env / in-repo default). */
  weightsPath?: string;
  /** Inject an ffmpeg decode for testing; defaults to the real binary. */
  decodeAudio?: (audio: Uint8Array, filename: string) => Promise<Float32Array>;
}

const TARGET_SAMPLE_RATE = 16_000;

function jsonResponse(
  status: number,
  payload: unknown,
): TranscriptionHttpResponse {
  return {
    status,
    headers: { 'Content-Type': 'application/json' },
    body: new TextEncoder().encode(JSON.stringify(payload)),
  };
}

function textResponse(
  status: number,
  text: string,
): TranscriptionHttpResponse {
  return {
    status,
    headers: { 'Content-Type': 'text/plain; charset=utf-8' },
    body: new TextEncoder().encode(text),
  };
}

function errorResponse(
  status: number,
  message: string,
): TranscriptionHttpResponse {
  return jsonResponse(status, {
    error: { message, type: 'invalid_request_error' },
  });
}

function headerValue(
  headers: Record<string, string | string[] | undefined>,
  name: string,
): string | undefined {
  const lower = name.toLowerCase();
  for (const key of Object.keys(headers)) {
    if (key.toLowerCase() === lower) {
      const value = headers[key];
      return Array.isArray(value) ? value[0] : value;
    }
  }
  return undefined;
}

// --- Default weights resolution -------------------------------------------

/** Resolve the in-repo whisper weights asset (q8 for accuracy). Walks up from
 *  this module to the monorepo root and points at wasm-modules/voice-stt. The
 *  .bin blobs are git-ignored (hosted on R2); env override takes precedence. */
function resolveDefaultWeightsPath(env: NodeJS.ProcessEnv): string {
  const fromEnv = env['WHISPER_STT_WEIGHTS'];
  if (fromEnv && fromEnv.trim()) return fromEnv.trim();

  // src/transcribe-audio.ts -> distributed-inference-host -> gnosis ->
  // open-source -> <repo root>. Four hops up reaches the monorepo root.
  const here = dirname(fileURLToPath(import.meta.url));
  const repoRoot = join(here, '..', '..', '..', '..');
  const assetsDir = join(repoRoot, 'wasm-modules', 'voice-stt', 'assets');

  // Prefer q8 (accuracy) over q4; fall back gracefully.
  for (const candidate of ['whisper_tiny_q8.bin', 'whisper_tiny_q4.bin']) {
    const full = join(assetsDir, candidate);
    if (existsSync(full)) return full;
  }
  return join(assetsDir, 'whisper_tiny_q8.bin');
}

// --- Lazy WhisperEngine singleton -----------------------------------------

let enginePromise: Promise<WhisperEngine> | null = null;
let loadedWeightsPath: string | null = null;

async function getEngine(weightsPath: string): Promise<WhisperEngine> {
  // Re-init if the requested weights differ from what was loaded.
  if (enginePromise && loadedWeightsPath === weightsPath) {
    return enginePromise;
  }
  loadedWeightsPath = weightsPath;
  enginePromise = (async () => {
    let stt: SttModule;
    try {
      stt = await import('@a0n/voice-stt');
    } catch (err) {
      throw new Error(
        `@a0n/voice-stt engine unavailable in this image (pkg-node wasm artifact missing): ${
          err instanceof Error ? err.message : err
        }`,
      );
    }
    if (!existsSync(weightsPath)) {
      throw new Error(
        `whisper weights not found at ${weightsPath}; set WHISPER_STT_WEIGHTS or host the .bin (R2)`,
      );
    }
    const weights = readFileSync(weightsPath);
    const engine = new stt.WhisperEngine();
    engine.load_model(new Uint8Array(weights));
    return engine;
  })();

  try {
    return await enginePromise;
  } catch (err) {
    // Allow a later request to retry after a transient failure.
    enginePromise = null;
    loadedWeightsPath = null;
    throw err;
  }
}

// --- Minimal multipart/form-data parser -----------------------------------

interface MultipartPart {
  name?: string;
  filename?: string;
  contentType?: string;
  data: Uint8Array;
}

function parseBoundary(contentType: string | undefined): string | null {
  if (!contentType) return null;
  const match = /boundary="?([^";]+)"?/i.exec(contentType);
  return match ? match[1]! : null;
}

function indexOfBytes(
  haystack: Uint8Array,
  needle: Uint8Array,
  from: number,
): number {
  outer: for (let i = from; i <= haystack.length - needle.length; i++) {
    for (let j = 0; j < needle.length; j++) {
      if (haystack[i + j] !== needle[j]) continue outer;
    }
    return i;
  }
  return -1;
}

/** Parse a multipart/form-data body into its parts. Boundary-driven byte scan
 *  so binary file payloads survive intact (no utf8 round-trip corruption). */
function parseMultipart(
  body: Uint8Array,
  boundary: string,
): MultipartPart[] {
  const enc = new TextEncoder();
  const delimiter = enc.encode(`--${boundary}`);
  const crlf = enc.encode('\r\n');
  const headerSep = enc.encode('\r\n\r\n');
  const parts: MultipartPart[] = [];

  let cursor = indexOfBytes(body, delimiter, 0);
  if (cursor < 0) return parts;

  while (cursor >= 0 && cursor < body.length) {
    let partStart = cursor + delimiter.length;
    // Closing delimiter is `--boundary--`.
    if (
      body[partStart] === 0x2d &&
      body[partStart + 1] === 0x2d
    ) {
      break;
    }
    // Skip the CRLF after the boundary line.
    if (
      body[partStart] === crlf[0] &&
      body[partStart + 1] === crlf[1]
    ) {
      partStart += crlf.length;
    }

    const headerEnd = indexOfBytes(body, headerSep, partStart);
    if (headerEnd < 0) break;
    const rawHeaders = new TextDecoder().decode(
      body.subarray(partStart, headerEnd),
    );
    const dataStart = headerEnd + headerSep.length;

    const nextBoundary = indexOfBytes(body, delimiter, dataStart);
    if (nextBoundary < 0) break;
    // Data ends just before the CRLF that precedes the next boundary.
    let dataEnd = nextBoundary;
    if (
      dataEnd >= 2 &&
      body[dataEnd - 2] === crlf[0] &&
      body[dataEnd - 1] === crlf[1]
    ) {
      dataEnd -= 2;
    }

    const part: MultipartPart = {
      data: body.subarray(dataStart, dataEnd),
    };
    const dispMatch = /content-disposition:[^\r\n]*/i.exec(rawHeaders);
    if (dispMatch) {
      const disp = dispMatch[0];
      const nameMatch = /name="([^"]*)"/i.exec(disp);
      if (nameMatch) part.name = nameMatch[1];
      const fileMatch = /filename="([^"]*)"/i.exec(disp);
      if (fileMatch) part.filename = fileMatch[1];
    }
    const ctMatch = /content-type:\s*([^\r\n]+)/i.exec(rawHeaders);
    if (ctMatch) part.contentType = ctMatch[1]!.trim();

    parts.push(part);
    cursor = nextBoundary;
  }

  return parts;
}

function fieldText(parts: MultipartPart[], name: string): string | undefined {
  const part = parts.find((p) => p.name === name && !p.filename);
  if (!part) return undefined;
  return new TextDecoder().decode(part.data);
}

// --- ffmpeg decode to 16 kHz mono f32 -------------------------------------

function runFfmpeg(args: string[]): Promise<{ ok: boolean; stderr: string }> {
  return new Promise((resolve) => {
    const child = spawn('ffmpeg', args, { stdio: ['ignore', 'ignore', 'pipe'] });
    let stderr = '';
    child.stderr?.on('data', (chunk: Buffer) => {
      stderr += chunk.toString('utf8');
    });
    const timer = setTimeout(() => child.kill('SIGTERM'), 60_000);
    child.on('close', (code) => {
      clearTimeout(timer);
      resolve({ ok: code === 0, stderr });
    });
    child.on('error', (err) => {
      clearTimeout(timer);
      resolve({ ok: false, stderr: err.message });
    });
  });
}

/** Decode arbitrary audio (wav/mp3/webm/m4a/...) to 16 kHz mono f32 PCM by
 *  shelling out to ffmpeg through a temp dir. */
async function decodeAudioWithFfmpeg(
  audio: Uint8Array,
  filename: string,
): Promise<Float32Array> {
  const dir = mkdtempSync(join(tmpdir(), 'thoth-stt-'));
  const safeName = filename.replace(/[^A-Za-z0-9._-]/g, '_') || 'input';
  const inPath = join(dir, safeName);
  const outPath = join(dir, 'audio.f32le');
  try {
    writeFileSync(inPath, audio);
    const result = await runFfmpeg([
      '-hide_banner',
      '-loglevel',
      'error',
      '-y',
      '-i',
      inPath,
      '-ar',
      String(TARGET_SAMPLE_RATE),
      '-ac',
      '1',
      '-f',
      'f32le',
      outPath,
    ]);
    if (!result.ok || !existsSync(outPath)) {
      throw new Error(
        `ffmpeg failed to decode audio: ${result.stderr.trim() || 'unknown error'}`,
      );
    }
    const raw = readFileSync(outPath);
    // readFileSync may hand back a Buffer whose byteOffset is non-zero or whose
    // length isn't a multiple of 4; copy into a fresh aligned buffer.
    const usableLen = raw.length - (raw.length % 4);
    const aligned = new Uint8Array(usableLen);
    aligned.set(raw.subarray(0, usableLen));
    return new Float32Array(aligned.buffer);
  } finally {
    rmSync(dir, { recursive: true, force: true });
  }
}

// --- Public handler --------------------------------------------------------

/**
 * Handles the Transcription Request request flow.
 */
export async function handleTranscriptionRequest(
  input: TranscriptionRequestInput,
  env: NodeJS.ProcessEnv = process.env,
): Promise<TranscriptionHttpResponse> {
  const contentType = headerValue(input.headers, 'content-type');
  const boundary = parseBoundary(contentType);
  if (!boundary) {
    return errorResponse(
      400,
      'expected multipart/form-data with a boundary (OpenAI Whisper API shape)',
    );
  }

  let parts: MultipartPart[];
  try {
    parts = parseMultipart(input.body, boundary);
  } catch (err) {
    return errorResponse(
      400,
      `failed to parse multipart body: ${err instanceof Error ? err.message : String(err)}`,
    );
  }

  const filePart = parts.find((p) => p.name === 'file' && p.filename != null);
  if (!filePart || filePart.data.length === 0) {
    return errorResponse(400, 'missing required `file` form field (audio bytes)');
  }

  const responseFormat = (
    fieldText(parts, 'response_format')?.trim().toLowerCase() ?? 'json'
  ) as TranscriptionResponseFormat;
  if (responseFormat !== 'json' && responseFormat !== 'text') {
    return errorResponse(
      400,
      'response_format must be "json" or "text"',
    );
  }
  const language = fieldText(parts, 'language')?.trim();

  // Decode to 16 kHz mono f32.
  let pcm: Float32Array;
  try {
    const decode = input.decodeAudio ?? decodeAudioWithFfmpeg;
    pcm = await decode(filePart.data, filePart.filename ?? 'audio');
  } catch (err) {
    return errorResponse(
      400,
      `audio decode failed: ${err instanceof Error ? err.message : String(err)}`,
    );
  }
  if (pcm.length === 0) {
    return errorResponse(400, 'decoded audio is empty');
  }

  // Run sovereign Whisper STT.
  let text: string;
  try {
    const weightsPath =
      input.weightsPath ?? resolveDefaultWeightsPath(env);
    const engine = await getEngine(weightsPath);
    if (language) {
      try {
        engine.set_language(language);
      } catch {
        // Non-fatal: fall back to auto-detect.
      }
    }
    const result = engine.transcribe(pcm, TARGET_SAMPLE_RATE) as {
      text?: unknown;
    };
    text = typeof result?.text === 'string' ? result.text.trim() : '';
  } catch (err) {
    return jsonResponse(500, {
      error: {
        message: `transcription failed: ${err instanceof Error ? err.message : String(err)}`,
        type: 'server_error',
      },
    });
  }

  if (responseFormat === 'text') {
    return textResponse(200, text);
  }
  return jsonResponse(200, { text });
}
