// @a0n/.ignored_gnosis/distributed-inference-host/src/transport-selector.ts
/**
 * transport-selector.ts — single entry point for opening an `IStation`
 * regardless of wire transport.
 *
 * The URL scheme picks the transport:
 *
 *   scheme               -> class               -> file
 *   ---------------------------------------------------------------------
 *   http://...           -> Station             -> station.ts
 *   https://...          -> Station             -> station.ts
 *   unix:///path/...     -> StationUds          -> station-uds.ts
 *   /path/to/socket      -> StationUds          -> (bare path, UDS)
 *   flow+unix:///path    -> StationFlow         -> station-flow.ts
 *   flow+udp://host:port -> StationFlowUdp      -> station-flow-udp.ts
 *   ws+https://host/...  -> WSStation           -> station-ws.ts (Death #3)
 *   ws+http://host/...   -> WSStation           -> station-ws.ts (Death #3)
 *
 * Why a selector
 * --------------
 * Pneuma's chain (and any future caller) used to hand-roll URL prefix
 * sniffing in `mesh-llm.ts` (`url.startsWith('/') || url.startsWith('unix://')`).
 * That works for two transports but explodes combinatorially when adding
 * aeon-flow UDS, aeon-flow UDP, and zero-copy UDS variants. This file owns
 * the dispatch table; callers depend ONLY on `IStation`.
 *
 * Diversity-theory framing
 * ------------------------
 * We deliberately keep multiple transports rather than crowning a winner:
 * HTTP wins for cross-host, UDS wins for local sub-millisecond floor,
 * aeon-flow wins for symmetric framing with the rest of the aeon ecosystem,
 * UDP wins when packet loss is acceptable for speculative paths. Each is a
 * different point in the (latency, reach, framing-cost) tradeoff space; the
 * selector lets the operator pick per-station, not per-deployment.
 *
 * Default
 * -------
 * The default transport for a `tri-g4-*` worker URL stays HTTPS. Operators
 * opt in to the WS substrate per-mesh by prefixing `ws+` to the existing
 * URL (e.g. `https://tri-g4-a-00...` → `ws+https://tri-g4-a-00...`). This
 * matches the staged-rollout pattern in
 * `docs/wsstation-dispatcher-wiring.md`.
 */

import type { IStation } from './types';
import { Station } from './station';
import { StationUds } from './station-uds';
import { StationFlow } from './station-flow';
import { StationFlowUdp } from './station-flow-udp';
import { WSStation, type WSStationPhase } from './station-ws';

export type StationRole = 'entry' | 'mid' | 'exit' | 'cobordism';

export interface OpenStationSpec {
  /** Wire address (see scheme table at top of file). */
  url: string;
  /** Inclusive layer-range start, exclusive end. */
  layerRange: [number, number];
  /** Optional declared role for routing. */
  role?: StationRole;
  /** Per-call timeout in ms — applied where the transport accepts it. */
  timeoutMs?: number;
  /**
   * Required when `url` uses `ws+http(s)://` — the trisplit phase this
   * station owns (split-a, split-b, or split-c). The /flow handler on the
   * deployed worker dispatches on (stream-id namespace, phase) to pick
   * the kernel op. Ignored for non-WS transports.
   */
  wsPhase?: WSStationPhase;
  /**
   * Optional Bearer token for the WS upgrade. Default `mesh-bench`.
   */
  wsBearerToken?: string;
  /**
   * Optional injected constructors. Tests pass stubs here so we can verify
   * scheme dispatch without spinning up real sockets / HTTP servers.
   * Production callers leave this undefined and get the real classes.
   */
  factories?: Partial<TransportFactories>;
}

export interface TransportFactories {
  http: (
    url: string,
    layerRange: [number, number],
    role: StationRole | undefined,
  ) => IStation;
  uds: (
    socketPath: string,
    layerRange: [number, number],
    role: StationRole | undefined,
    timeoutMs: number | undefined,
  ) => IStation;
  flowUds: (
    socketPath: string,
    layerRange: [number, number],
    role: StationRole | undefined,
    timeoutMs: number | undefined,
  ) => IStation;
  flowUdp: (
    host: string,
    port: number,
    layerRange: [number, number],
    role: StationRole | undefined,
    timeoutMs: number | undefined,
  ) => IStation;
  ws: (
    url: string,
    layerRange: [number, number],
    role: StationRole | undefined,
    phase: WSStationPhase,
    bearerToken: string | undefined,
  ) => IStation;
}

const DEFAULT_FACTORIES: TransportFactories = {
  http: (url, [startLayer, endLayer], role) =>
    new Station({ endpoint: url, startLayer, endLayer, role }),
  uds: (socketPath, [startLayer, endLayer], role, timeoutMs) =>
    new StationUds({ socketPath, startLayer, endLayer, role, timeoutMs }),
  flowUds: (socketPath, [startLayer, endLayer], role, timeoutMs) =>
    new StationFlow({ socketPath, startLayer, endLayer, role, timeoutMs }),
  flowUdp: (host, port, [startLayer, endLayer], role, timeoutMs) =>
    new StationFlowUdp({ host, port, startLayer, endLayer, role, timeoutMs }),
  ws: (url, [startLayer, endLayer], role, phase, bearerToken) =>
    new WSStation({
      endpoint: url,
      startLayer,
      endLayer,
      role,
      phase,
      bearerToken,
    }),
};

/**
 * Dispatch a `spec.url` to the right transport class and return an
 * `IStation` the Pipeline can drive without caring which wire it picked.
 *
 * Throws on unknown / malformed schemes — the caller mis-typed the URL.
 */
export function openStation(spec: OpenStationSpec): IStation {
  const factories: TransportFactories = {
    ...DEFAULT_FACTORIES,
    ...(spec.factories ?? {}),
  };
  const { url, layerRange, role, timeoutMs } = spec;

  // ws+http(s):// — Death #3 persistent WebSocket substrate. Must precede
  // the bare http:// match so `ws+http://...` is not misrouted.
  if (url.startsWith('ws+https://') || url.startsWith('ws+http://')) {
    if (!spec.wsPhase) {
      throw new Error(
        `openStation: ws+http(s):// requires wsPhase ('split-a' | 'split-b' | 'split-c'); got url='${url}'`,
      );
    }
    const innerUrl = url.startsWith('ws+https://')
      ? 'https://' + url.slice('ws+https://'.length)
      : 'http://' + url.slice('ws+http://'.length);
    return factories.ws(innerUrl, layerRange, role, spec.wsPhase, spec.wsBearerToken);
  }

  if (url.startsWith('http://') || url.startsWith('https://')) {
    return factories.http(url, layerRange, role);
  }

  if (url.startsWith('flow+unix://')) {
    const socketPath = url.slice('flow+unix://'.length);
    if (!socketPath) {
      throw new Error(`openStation: flow+unix:// requires a socket path (got '${url}')`);
    }
    return factories.flowUds(socketPath, layerRange, role, timeoutMs);
  }

  if (url.startsWith('flow+udp://')) {
    const rest = url.slice('flow+udp://'.length);
    const colon = rest.lastIndexOf(':');
    if (colon < 0) {
      throw new Error(
        `openStation: flow+udp:// requires host:port (got '${url}')`,
      );
    }
    const host = rest.slice(0, colon);
    const port = Number.parseInt(rest.slice(colon + 1), 10);
    if (!host || !Number.isFinite(port) || port <= 0 || port > 65535) {
      throw new Error(
        `openStation: flow+udp:// host/port malformed (host='${host}', port='${rest.slice(colon + 1)}')`,
      );
    }
    return factories.flowUdp(host, port, layerRange, role, timeoutMs);
  }

  if (url.startsWith('unix://')) {
    const socketPath = url.slice('unix://'.length);
    if (!socketPath) {
      throw new Error(`openStation: unix:// requires a socket path (got '${url}')`);
    }
    return factories.uds(socketPath, layerRange, role, timeoutMs);
  }

  if (url.startsWith('/')) {
    // Bare absolute path — treat as UDS, matching the legacy mesh-llm.ts
    // behavior. Consider migrating callers to the explicit `unix://` form.
    return factories.uds(url, layerRange, role, timeoutMs);
  }

  throw new Error(
    `openStation: unrecognized URL scheme in '${url}'. ` +
      `Expected one of: http(s)://, unix://, flow+unix://, flow+udp://, or a bare /absolute/path.`,
  );
}
