// @a0n/.ignored_gnosis/distributed-inference-host/src/unlimited-ocr-qspec.ts
export const UNLIMITED_OCR_MODEL_ID = 'unlimited-ocr';
export const UNLIMITED_OCR_REPO_ID = 'sahilchachra/Unlimited-OCR-GGUF';
export const UNLIMITED_OCR_DEFAULT_QUANT = 'Q4_K_M';
export const UNLIMITED_OCR_MODEL_FILENAME = 'Unlimited-OCR-Q4_K_M.gguf';
export const UNLIMITED_OCR_MMPROJ_FILENAME = 'mmproj-Unlimited-OCR-F16.gguf';
export const UNLIMITED_OCR_CHAT_TEMPLATE = 'deepseek-ocr';
export const UNLIMITED_OCR_DEFAULT_PROMPT =
  '<|grounding|>Convert the document to markdown.';
export const UNLIMITED_OCR_QSPEC_MAX_TOKENS = 512;

export interface UnlimitedOcrArtifactSpec {
  repoId: string;
  quant: string;
  modelFilename: string;
  mmprojFilename: string;
  chatTemplate: typeof UNLIMITED_OCR_CHAT_TEMPLATE;
}

export interface UnlimitedOcrProbeSpec {
  probe: 'unlimited-ocr-gguf';
  model: typeof UNLIMITED_OCR_MODEL_ID;
  modality: 'ocr';
  gate_tier: 'visible-document-ocr';
  artifacts: UnlimitedOcrArtifactSpec;
  prompt: string;
  input: {
    pdf?: string;
    image?: string;
    renderPage: number;
    renderDpi: number;
  };
  decode: {
    maxTokens: number;
    ignoreEos: boolean;
  };
  expect: {
    minTextChars: number;
    requiredTerms: readonly string[];
  };
}

export interface UnlimitedOcrChatContentPart {
  type: 'text' | 'image_url';
  text?: string;
  image_url?: { url: string };
}

export interface UnlimitedOcrChatCompletionRequest {
  model: string;
  temperature: 0;
  max_tokens: number;
  messages: Array<{
    role: 'user';
    content: UnlimitedOcrChatContentPart[];
  }>;
}

export const UNLIMITED_OCR_ARTIFACTS: UnlimitedOcrArtifactSpec = {
  repoId: UNLIMITED_OCR_REPO_ID,
  quant: UNLIMITED_OCR_DEFAULT_QUANT,
  modelFilename: UNLIMITED_OCR_MODEL_FILENAME,
  mmprojFilename: UNLIMITED_OCR_MMPROJ_FILENAME,
  chatTemplate: UNLIMITED_OCR_CHAT_TEMPLATE,
};

/**
 * Builds the Unlimited Ocr Chat Request.
 */
export function buildUnlimitedOcrChatRequest(options: {
  imageUrl: string;
  prompt?: string;
  maxTokens?: number;
  model?: string;
}): UnlimitedOcrChatCompletionRequest {
  const prompt = options.prompt ?? UNLIMITED_OCR_DEFAULT_PROMPT;
  const maxTokens = options.maxTokens ?? 4096;
  const model = options.model ?? UNLIMITED_OCR_MODEL_ID;

  if (!options.imageUrl.trim()) {
    throw new Error('Unlimited OCR request requires an image URL');
  }
  if (!prompt.trim()) {
    throw new Error('Unlimited OCR request requires a non-empty prompt');
  }
  if (!Number.isInteger(maxTokens) || maxTokens <= 0) {
    throw new Error('Unlimited OCR request requires a positive max token count');
  }

  return {
    model,
    temperature: 0,
    max_tokens: maxTokens,
    messages: [
      {
        role: 'user',
        content: [
          { type: 'text', text: prompt },
          { type: 'image_url', image_url: { url: options.imageUrl } },
        ],
      },
    ],
  };
}

/**
 * Handles the gnosis validate Unlimited Ocr Output workflow.
 */
export function validateUnlimitedOcrOutput(
  output: string,
  spec: Pick<UnlimitedOcrProbeSpec, 'expect'>
): { pass: true } | { pass: false; missingTerms: string[]; textChars: number } {
  const normalized = output.toLowerCase();
  const missingTerms = spec.expect.requiredTerms.filter(
    (term) => !normalized.includes(term.toLowerCase())
  );
  const textChars = output.trim().length;
  if (textChars >= spec.expect.minTextChars && missingTerms.length === 0) {
    return { pass: true };
  }
  return { pass: false, missingTerms, textChars };
}
