// @a0n/.ignored_gnosis/distributed-inference-host/src/unlimited-ocr-runner.ts
import { execFile as execFileCallback } from 'node:child_process';
import { existsSync } from 'node:fs';
import { mkdir, writeFile } from 'node:fs/promises';
import { basename, extname, join, resolve } from 'node:path';
import { promisify } from 'node:util';

import {
  UNLIMITED_OCR_CHAT_TEMPLATE,
  UNLIMITED_OCR_DEFAULT_PROMPT,
  UNLIMITED_OCR_MMPROJ_FILENAME,
  UNLIMITED_OCR_MODEL_FILENAME,
  UNLIMITED_OCR_MODEL_ID,
  UNLIMITED_OCR_QSPEC_MAX_TOKENS,
  validateUnlimitedOcrOutput,
} from './unlimited-ocr-qspec';

const execFile = promisify(execFileCallback);
const DEFAULT_MODEL_DIR = '/tmp/unlimited-ocr-gguf';
const DEFAULT_OUT_DIR = '/tmp/unlimited-ocr-qspec';
const DEFAULT_TIMEOUT_MS = 20 * 60 * 1000;
const DEFAULT_MAX_BUFFER = 32 * 1024 * 1024;

export type UnlimitedOcrDocumentKind = 'pdf' | 'image';

export interface UnlimitedOcrCommandRun {
  readonly command: string;
  readonly args: readonly string[];
  readonly cwd?: string;
  readonly timeoutMs?: number;
  readonly maxBuffer?: number;
}

export interface UnlimitedOcrCommandOutput {
  readonly stdout: string;
  readonly stderr: string;
}

export type UnlimitedOcrCommandRunner = (
  run: UnlimitedOcrCommandRun
) => Promise<UnlimitedOcrCommandOutput>;

export interface UnlimitedOcrDocumentInput {
  readonly sourcePath: string;
  readonly kind?: UnlimitedOcrDocumentKind;
  readonly sourceId?: string;
  readonly prompt?: string;
  readonly renderPage?: number;
  readonly renderDpi?: number;
  readonly maxTokens?: number;
  readonly ignoreEos?: boolean;
  readonly requiredTerms?: readonly string[];
  readonly minTextChars?: number;
}

export interface UnlimitedOcrRunnerOptions {
  readonly modelDir?: string;
  readonly outDir?: string;
  readonly llamaMtmdCli?: string;
  readonly pdftoppmCli?: string;
  readonly timeoutMs?: number;
  readonly maxBuffer?: number;
  readonly commandRunner?: UnlimitedOcrCommandRunner;
  readonly skipInference?: boolean;
}

export interface UnlimitedOcrDocumentResult {
  readonly model: typeof UNLIMITED_OCR_MODEL_ID;
  readonly sourcePath: string;
  readonly sourceId?: string;
  readonly imagePath: string;
  readonly outputPath?: string;
  readonly text: string;
  readonly prompt: string;
  readonly renderPage: number;
  readonly renderDpi: number;
  readonly maxTokens: number;
  readonly ignoreEos: boolean;
  readonly modelPath: string;
  readonly mmprojPath: string;
  readonly validation:
    | { readonly pass: true }
    | { readonly pass: false; readonly missingTerms: string[]; readonly textChars: number };
  readonly commands: readonly UnlimitedOcrCommandRun[];
  readonly inference: 'ran' | 'skipped';
}

/**
 * Handles the gnosis run Unlimited Ocr Document workflow.
 */
export async function runUnlimitedOcrDocument(
  input: UnlimitedOcrDocumentInput,
  options: UnlimitedOcrRunnerOptions = {}
): Promise<UnlimitedOcrDocumentResult> {
  const sourcePath = resolve(input.sourcePath);
  if (!existsSync(sourcePath)) {
    throw new Error(`Unlimited OCR source not found: ${sourcePath}`);
  }

  const kind = inferDocumentKind(sourcePath, input.kind);
  const outDir = options.outDir ?? DEFAULT_OUT_DIR;
  const modelDir = options.modelDir ?? DEFAULT_MODEL_DIR;
  const runner = options.commandRunner ?? defaultCommandRunner;
  const commands: UnlimitedOcrCommandRun[] = [];
  await mkdir(outDir, { recursive: true });

  const renderPage = positiveInteger(input.renderPage ?? 1, 'renderPage');
  const renderDpi = positiveInteger(input.renderDpi ?? 180, 'renderDpi');
  const maxTokens = positiveInteger(
    input.maxTokens ?? UNLIMITED_OCR_QSPEC_MAX_TOKENS,
    'maxTokens'
  );
  const prompt = input.prompt ?? UNLIMITED_OCR_DEFAULT_PROMPT;
  if (prompt.trim().length === 0) {
    throw new Error('Unlimited OCR prompt must be non-empty.');
  }

  const imagePath =
    kind === 'pdf'
      ? await renderPdfPage({
          sourcePath,
          outDir,
          renderPage,
          renderDpi,
          pdftoppmCli: options.pdftoppmCli ?? 'pdftoppm',
          runner,
          commands,
          timeoutMs: options.timeoutMs ?? DEFAULT_TIMEOUT_MS,
          maxBuffer: options.maxBuffer ?? DEFAULT_MAX_BUFFER,
        })
      : sourcePath;

  const modelPath = join(modelDir, UNLIMITED_OCR_MODEL_FILENAME);
  const mmprojPath = join(modelDir, UNLIMITED_OCR_MMPROJ_FILENAME);
  const requiredTerms = input.requiredTerms ?? [];
  const minTextChars = input.minTextChars ?? 80;

  if (options.skipInference) {
    return {
      model: UNLIMITED_OCR_MODEL_ID,
      ...(input.sourceId ? { sourceId: input.sourceId } : {}),
      sourcePath,
      imagePath,
      text: '',
      prompt,
      renderPage,
      renderDpi,
      maxTokens,
      ignoreEos: input.ignoreEos ?? true,
      modelPath,
      mmprojPath,
      validation: validateUnlimitedOcrOutput('', {
        expect: { minTextChars, requiredTerms },
      }),
      commands,
      inference: 'skipped',
    };
  }

  if (!existsSync(modelPath) || !existsSync(mmprojPath)) {
    throw new Error(`missing Unlimited OCR GGUF artifacts in ${modelDir}`);
  }

  const llamaArgs = [
    '-m',
    modelPath,
    '--mmproj',
    mmprojPath,
    '--image',
    imagePath,
    '--chat-template',
    UNLIMITED_OCR_CHAT_TEMPLATE,
    '--temp',
    '0',
    '-n',
    String(maxTokens),
    '-p',
    prompt,
  ];
  const ignoreEos = input.ignoreEos ?? true;
  if (ignoreEos) {
    llamaArgs.push('--ignore-eos');
  }

  const llamaRun: UnlimitedOcrCommandRun = {
    command: options.llamaMtmdCli ?? 'llama-mtmd-cli',
    args: llamaArgs,
    timeoutMs: options.timeoutMs ?? DEFAULT_TIMEOUT_MS,
    maxBuffer: options.maxBuffer ?? DEFAULT_MAX_BUFFER,
  };
  commands.push(llamaRun);
  const { stdout } = await runner(llamaRun);
  const outputPath = join(outDir, 'unlimited-ocr-output.md');
  await writeFile(outputPath, stdout);

  return {
    model: UNLIMITED_OCR_MODEL_ID,
    ...(input.sourceId ? { sourceId: input.sourceId } : {}),
    sourcePath,
    imagePath,
    outputPath,
    text: stdout,
    prompt,
    renderPage,
    renderDpi,
    maxTokens,
    ignoreEos,
    modelPath,
    mmprojPath,
    validation: validateUnlimitedOcrOutput(stdout, {
      expect: { minTextChars, requiredTerms },
    }),
    commands,
    inference: 'ran',
  };
}

/**
 * Handles the gnosis infer Document Kind workflow.
 */
export function inferDocumentKind(
  sourcePath: string,
  explicitKind?: UnlimitedOcrDocumentKind
): UnlimitedOcrDocumentKind {
  if (explicitKind) {
    return explicitKind;
  }
  const extension = extname(sourcePath).toLowerCase();
  if (extension === '.pdf') {
    return 'pdf';
  }
  if (['.png', '.jpg', '.jpeg', '.webp'].includes(extension)) {
    return 'image';
  }
  throw new Error(`Unsupported Unlimited OCR source type: ${extension || '(none)'}`);
}

async function renderPdfPage(input: {
  readonly sourcePath: string;
  readonly outDir: string;
  readonly renderPage: number;
  readonly renderDpi: number;
  readonly pdftoppmCli: string;
  readonly runner: UnlimitedOcrCommandRunner;
  readonly commands: UnlimitedOcrCommandRun[];
  readonly timeoutMs: number;
  readonly maxBuffer: number;
}): Promise<string> {
  const prefix = join(
    input.outDir,
    basename(input.sourcePath, '.pdf').replace(/[^A-Za-z0-9._-]+/g, '-')
  );
  const run: UnlimitedOcrCommandRun = {
    command: input.pdftoppmCli,
    args: [
      '-png',
      '-f',
      String(input.renderPage),
      '-singlefile',
      '-r',
      String(input.renderDpi),
      input.sourcePath,
      prefix,
    ],
    timeoutMs: input.timeoutMs,
    maxBuffer: input.maxBuffer,
  };
  input.commands.push(run);
  await input.runner(run);
  const imagePath = `${prefix}.png`;
  if (!existsSync(imagePath)) {
    throw new Error(`rendered Unlimited OCR image missing: ${imagePath}`);
  }
  return imagePath;
}

function positiveInteger(value: number, name: string): number {
  if (!Number.isInteger(value) || value <= 0) {
    throw new Error(`Unlimited OCR ${name} must be a positive integer.`);
  }
  return value;
}

async function defaultCommandRunner(
  run: UnlimitedOcrCommandRun
): Promise<UnlimitedOcrCommandOutput> {
  const output = await execFile(run.command, [...run.args], {
    ...(run.cwd ? { cwd: run.cwd } : {}),
    encoding: 'utf8',
    timeout: run.timeoutMs ?? DEFAULT_TIMEOUT_MS,
    maxBuffer: run.maxBuffer ?? DEFAULT_MAX_BUFFER,
  });
  return {
    stdout: output.stdout,
    stderr: output.stderr,
  };
}
