// @a0n/.ignored_gnosis/distributed-inference-host/src/video-deblur.ts
// End-to-end video deblur orchestrator. The φ-attractor σ-selection from
// `deblurAnyImage` is run ONCE on frame 0 and the resolved σ is then locked
// across every subsequent frame — running 'auto' per-frame is far too
// expensive for video, and blur typically does not change frame-to-frame
// inside a single clip, so amortising the search over the whole stream is
// both cheaper and stable. ffmpeg is invoked directly via `node:child_process`
// (no `fluent-ffmpeg` dep — the dep tree stays clean). Workers complete out of
// order, so a sequence-numbered reorder buffer holds early arrivals until
// their predecessors land and the encoder receives strictly in-order frames.

import { spawn, type ChildProcess } from 'node:child_process';
import { Worker } from 'node:worker_threads';
import { resolve as resolvePath, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import os from 'node:os';
import { deblurAnyImage, type DeblurAnyImageOptions } from './deblur-any-image';
import { floatPlaneToUint8Clamped } from './invisible-watermark';

export interface DeblurVideoArgs {
  /** Input video path (anything ffmpeg can decode). */
  input: string;
  /** Output video path. .mp4 default-encoded as H.264. */
  output: string;
  /** Number of worker_threads. Default `max(1, floor(os.cpus().length / 2))`.
   *  Pass `0` to bypass workers and process frames inline on the main thread —
   *  useful when running with a pure-TS source (vitest, dev) without a
   *  loader for worker_threads. Production with a built JS bundle should use
   *  the default. */
  workers?: number;
  /** Forwarded to `deblurAnyImage` per frame. `sigma` is resolved once on
   *  frame 0 then locked across the rest of the video. */
  options?: DeblurAnyImageOptions;
  /** Optional progress callback. Called every `progressInterval` frames (default 8). */
  onProgress?: (info: { current: number; total?: number; fps: number; elapsedMs: number }) => void;
  progressInterval?: number;
  /** ffmpeg / ffprobe binary names. Default 'ffmpeg' and 'ffprobe' (system PATH). */
  ffmpegPath?: string;
  ffprobePath?: string;
}

export interface DeblurVideoResult {
  framesProcessed: number;
  selectedSigma: number;
  width: number;
  height: number;
  fps: number;
  elapsedMs: number;
  averageFps: number;
}

interface ProbedMeta {
  width: number;
  height: number;
  fps: number;
  nbFrames?: number;
}

function parseFps(rate: string): number {
  // ffprobe gives strings like "30000/1001" or "30/1" or "25/1".
  if (typeof rate !== 'string' || rate.length === 0) return 0;
  const parts = rate.split('/');
  if (parts.length === 1) {
    const v = Number(parts[0]);
    return Number.isFinite(v) ? v : 0;
  }
  const num = Number(parts[0]);
  const den = Number(parts[1]);
  if (!Number.isFinite(num) || !Number.isFinite(den) || den === 0) return 0;
  return num / den;
}

async function probeVideo(input: string, ffprobePath: string): Promise<ProbedMeta> {
  return new Promise<ProbedMeta>((resolve, reject) => {
    const proc = spawn(
      ffprobePath,
      [
        '-v', 'error',
        '-select_streams', 'v:0',
        '-show_entries', 'stream=width,height,r_frame_rate,nb_frames',
        '-of', 'json',
        input,
      ],
      { stdio: ['ignore', 'pipe', 'pipe'] },
    );

    const outChunks: Buffer[] = [];
    const errChunks: Buffer[] = [];
    proc.stdout.on('data', (c: Buffer) => outChunks.push(c));
    proc.stderr.on('data', (c: Buffer) => errChunks.push(c));
    proc.on('error', (err) => reject(err));
    proc.on('close', (code) => {
      if (code !== 0) {
        const stderr = Buffer.concat(errChunks).toString('utf8');
        reject(new Error(`ffprobe exited with code ${code}: ${stderr}`));
        return;
      }
      try {
        const json = JSON.parse(Buffer.concat(outChunks).toString('utf8'));
        const stream = Array.isArray(json.streams) ? json.streams[0] : undefined;
        if (!stream) {
          reject(new Error('ffprobe: no video stream found'));
          return;
        }
        const width = Number(stream.width);
        const height = Number(stream.height);
        const fps = parseFps(String(stream.r_frame_rate ?? ''));
        const nbRaw = stream.nb_frames;
        let nbFrames: number | undefined;
        if (nbRaw !== undefined && nbRaw !== null && nbRaw !== 'N/A') {
          const n = Number(nbRaw);
          if (Number.isFinite(n) && n > 0) nbFrames = n;
        }
        if (!Number.isFinite(width) || !Number.isFinite(height) || width <= 0 || height <= 0) {
          reject(new Error(`ffprobe: invalid dimensions ${width}x${height}`));
          return;
        }
        if (!Number.isFinite(fps) || fps <= 0) {
          reject(new Error(`ffprobe: invalid fps from r_frame_rate=${stream.r_frame_rate}`));
          return;
        }
        resolve({ width, height, fps, nbFrames });
      } catch (err) {
        reject(new Error(`ffprobe parse error: ${(err as Error).message}`));
      }
    });
  });
}

/**
 * Frame splitter. Wraps a readable byte stream and emits fixed-size frames
 * (width * height * 3 bytes). `next()` returns the next frame as a Buffer or
 * `null` when the stream has ended and no more frames are available.
 */
class FrameSplitter {
  private buffers: Buffer[] = [];
  private buffered = 0;
  private ended = false;
  private waiters: Array<(b: Buffer | null) => void> = [];
  private readonly frameBytes: number;

  constructor(
    stream: NodeJS.ReadableStream,
    frameBytes: number,
    private readonly onError: (err: Error) => void,
  ) {
    this.frameBytes = frameBytes;
    stream.on('data', (chunk: Buffer) => {
      this.buffers.push(chunk);
      this.buffered += chunk.length;
      this.flushWaiters();
    });
    stream.on('end', () => {
      this.ended = true;
      this.flushWaiters();
    });
    stream.on('error', (err: Error) => {
      this.ended = true;
      this.onError(err);
      this.flushWaiters();
    });
  }

  private takeFrame(): Buffer | null {
    if (this.buffered < this.frameBytes) return null;
    // Coalesce just enough bytes.
    const merged = Buffer.concat(this.buffers, this.buffered);
    const frame = merged.subarray(0, this.frameBytes);
    const rest = merged.subarray(this.frameBytes);
    this.buffers = rest.length > 0 ? [rest] : [];
    this.buffered = rest.length;
    // Copy so the caller can transfer the underlying ArrayBuffer without
    // also detaching the leftover bytes we still hold.
    return Buffer.from(frame);
  }

  private flushWaiters(): void {
    while (this.waiters.length > 0) {
      const f = this.takeFrame();
      if (f !== null) {
        const w = this.waiters.shift()!;
        w(f);
        continue;
      }
      if (this.ended) {
        const w = this.waiters.shift()!;
        w(null);
        continue;
      }
      return;
    }
  }

  next(): Promise<Buffer | null> {
    const f = this.takeFrame();
    if (f !== null) return Promise.resolve(f);
    if (this.ended) return Promise.resolve(null);
    return new Promise<Buffer | null>((resolve) => {
      this.waiters.push(resolve);
    });
  }
}

interface PoolWorker {
  worker: Worker;
  ready: Promise<void>;
  busy: boolean;
}

/** Convert a Node Buffer into a standalone ArrayBuffer suitable for transfer.
 *  We always copy because Buffer instances share an underlying pool, and
 *  transferring the pool would detach unrelated buffers. */
function bufferToTransferableArrayBuffer(b: Buffer): ArrayBuffer {
  const ab = new ArrayBuffer(b.length);
  new Uint8Array(ab).set(b);
  return ab;
}

/**
 * Handles the gnosis deblur Video workflow.
 */
export async function deblurVideo(args: DeblurVideoArgs): Promise<DeblurVideoResult> {
  if (typeof args.input !== 'string' || args.input.length === 0) {
    throw new RangeError('deblurVideo: args.input must be a non-empty string');
  }
  if (typeof args.output !== 'string' || args.output.length === 0) {
    throw new RangeError('deblurVideo: args.output must be a non-empty string');
  }

  const cpuCount = Math.max(1, os.cpus().length);
  const defaultWorkers = Math.max(1, Math.floor(cpuCount / 2));
  // inFlightCap below uses workers*2; in inline mode that becomes 0, which
  // would block the dispatcher entirely. Floor it at 1 in inline mode.
  void cpuCount; // referenced below
  let workerCount = args.workers ?? defaultWorkers;
  if (
    !Number.isFinite(workerCount)
    || !Number.isInteger(workerCount)
    || workerCount < 0
    || workerCount > cpuCount * 2
  ) {
    throw new RangeError(
      `deblurVideo: workers must be an integer in [0, ${cpuCount * 2}], got ${args.workers}`,
    );
  }
  const inlineMode = workerCount === 0;

  const progressInterval = args.progressInterval ?? 8;
  if (
    !Number.isFinite(progressInterval)
    || !Number.isInteger(progressInterval)
    || progressInterval < 1
  ) {
    throw new RangeError(`deblurVideo: progressInterval must be a positive integer, got ${args.progressInterval}`);
  }

  const ffmpegPath = args.ffmpegPath ?? 'ffmpeg';
  const ffprobePath = args.ffprobePath ?? 'ffprobe';

  // 1. Probe.
  const meta = await probeVideo(args.input, ffprobePath);
  const { width, height, fps, nbFrames } = meta;
  const frameBytes = width * height * 3;

  // 2. Spawn decoder.
  const decoder: ChildProcess = spawn(
    ffmpegPath,
    ['-i', args.input, '-f', 'rawvideo', '-pix_fmt', 'rgb24', '-'],
    { stdio: ['ignore', 'pipe', 'pipe'] },
  );
  const decoderStderr: Buffer[] = [];
  decoder.stderr!.on('data', (c: Buffer) => decoderStderr.push(c));

  // 3. Spawn encoder.
  const encoder: ChildProcess = spawn(
    ffmpegPath,
    [
      '-y',
      '-f', 'rawvideo',
      '-pix_fmt', 'rgb24',
      '-s', `${width}x${height}`,
      '-r', String(fps),
      '-i', '-',
      '-c:v', 'libx264',
      '-pix_fmt', 'yuv420p',
      '-preset', 'fast',
      args.output,
    ],
    { stdio: ['pipe', 'pipe', 'pipe'] },
  );
  const encoderStderr: Buffer[] = [];
  encoder.stderr!.on('data', (c: Buffer) => encoderStderr.push(c));

  // Cleanup helpers shared by the success and failure paths.
  const workersPool: PoolWorker[] = [];
  let cleaned = false;
  const cleanup = (): void => {
    if (cleaned) return;
    cleaned = true;
    for (const w of workersPool) {
      try { w.worker.terminate(); } catch { /* swallow */ }
    }
    try { decoder.kill('SIGKILL'); } catch { /* swallow */ }
    try {
      if (encoder.stdin && !encoder.stdin.destroyed) encoder.stdin.end();
    } catch { /* swallow */ }
    try { encoder.kill('SIGKILL'); } catch { /* swallow */ }
  };

  return new Promise<DeblurVideoResult>((resolve, reject) => {
    let settled = false;
    const fail = (err: Error): void => {
      if (settled) return;
      settled = true;
      cleanup();
      reject(err);
    };

    decoder.on('error', (err) => fail(new Error(`ffmpeg decoder error: ${err.message}`)));
    encoder.on('error', (err) => fail(new Error(`ffmpeg encoder error: ${err.message}`)));
    decoder.on('close', (code, signal) => {
      if (settled) return;
      if (code !== null && code !== 0) {
        const stderr = Buffer.concat(decoderStderr).toString('utf8');
        fail(new Error(`ffmpeg decoder exited with code ${code} signal=${signal}: ${stderr}`));
      }
    });

    (async () => {
      try {
        // 4. Frame splitter over decoder.stdout.
        const splitter = new FrameSplitter(decoder.stdout!, frameBytes, (err) => {
          fail(new Error(`decoder stdout error: ${err.message}`));
        });

        // 5. Read frame 0 and resolve σ on the main thread.
        const frame0 = await splitter.next();
        if (frame0 === null) {
          throw new Error('deblurVideo: decoder produced no frames');
        }

        const baseOptions: DeblurAnyImageOptions = { ...(args.options ?? {}) };
        const frame0Result = await deblurAnyImage(
          { data: new Uint8Array(frame0), width, height, channels: 3 },
          { ...baseOptions, watermarkFrameIndex: 0 },
        );
        const resolvedSigma = frame0Result.selectedSigma;

        // Lock σ for all subsequent frames.
        const lockedOptions: DeblurAnyImageOptions = { ...baseOptions, sigma: resolvedSigma };

        // 6. Spawn worker pool (skipped in inline mode).
        if (!inlineMode) {
          const here = dirname(fileURLToPath(import.meta.url));
          const workerPath = resolvePath(here, './video-frame-worker.ts');

          for (let i = 0; i < workerCount; i++) {
            const w = new Worker(workerPath);
            const ready = new Promise<void>((resReady, rejReady) => {
              const onMsg = (msg: any): void => {
                if (msg && msg.type === 'ready') {
                  w.off('message', onMsg);
                  w.off('error', onErr);
                  resReady();
                }
              };
              const onErr = (err: Error): void => {
                w.off('message', onMsg);
                w.off('error', onErr);
                rejReady(err);
              };
              w.on('message', onMsg);
              w.on('error', onErr);
            });
            workersPool.push({ worker: w, ready, busy: false });
          }

          await Promise.all(workersPool.map((w) => w.ready));

          for (const w of workersPool) {
            w.worker.postMessage({ type: 'init', width, height, options: lockedOptions });
            w.worker.on('error', (err) => fail(new Error(`worker error: ${err.message}`)));
          }
        }

        // 7. Reorder + dispatch state.
        const pending = new Map<number, { recovered: ArrayBuffer; selectedSigma: number }>();
        let nextOutSeq = 0;
        let framesWritten = 0;
        let inFlight = 0;
        const inFlightCap = inlineMode ? 1 : workerCount * 2;
        const startedAt = Date.now();
        let finalSigma = resolvedSigma;

        // Bookkeeping for a frame 0 fast-path: convert frame0Result.recovered
        // (Float32) into a uint8 buffer to feed the encoder, treating it as
        // seq=0 so the reorder logic remains uniform.
        const f0u8 = floatPlaneToUint8Clamped(frame0Result.recovered);
        pending.set(0, { recovered: toOwnedArrayBuffer(f0u8), selectedSigma: resolvedSigma });

        let allDispatched = false;
        let resolveAllDrained: (() => void) | null = null;
        const allDrained = new Promise<void>((res) => {
          resolveAllDrained = res;
        });

        const drainPending = async (): Promise<void> => {
          while (pending.has(nextOutSeq)) {
            const item = pending.get(nextOutSeq)!;
            pending.delete(nextOutSeq);
            const buf = Buffer.from(item.recovered);
            const ok = encoder.stdin!.write(buf);
            if (!ok) {
              await new Promise<void>((res) => encoder.stdin!.once('drain', () => res()));
            }
            framesWritten++;
            nextOutSeq++;
            finalSigma = item.selectedSigma;

            if (args.onProgress && framesWritten % progressInterval === 0) {
              const elapsedMs = Date.now() - startedAt;
              const elapsedSec = elapsedMs / 1000;
              const fpsNow = elapsedSec > 0 ? framesWritten / elapsedSec : 0;
              try {
                args.onProgress({
                  current: framesWritten,
                  total: nbFrames,
                  fps: fpsNow,
                  elapsedMs,
                });
              } catch { /* user callback errors must not abort the run */ }
            }
          }
          if (allDispatched && inFlight === 0 && pending.size === 0 && resolveAllDrained) {
            resolveAllDrained();
            resolveAllDrained = null;
          }
        };

        // Wire the persistent message + error handlers for each worker.
        for (const w of workersPool) {
          w.worker.on('message', (msg: any) => {
            if (!msg || typeof msg !== 'object') return;
            if (msg.type === 'done') {
              inFlight--;
              w.busy = false;
              pending.set(msg.seq, {
                recovered: msg.recovered as ArrayBuffer,
                selectedSigma: typeof msg.selectedSigma === 'number'
                  ? msg.selectedSigma
                  : resolvedSigma,
              });
              drainPending().catch((err: Error) => fail(err));
              dispatchLoop().catch((err: Error) => fail(err));
              return;
            }
            if (msg.type === 'error') {
              fail(new Error(`worker frame seq=${msg.seq} error: ${msg.error}`));
            }
          });
        }

        const findIdleWorker = (): PoolWorker | null => {
          for (const w of workersPool) {
            if (!w.busy) return w;
          }
          return null;
        };

        let nextInSeq = 1; // seq 0 was already handled on main thread.
        let readingDone = false;

        // Inline-mode helper: process a frame on the main thread.
        const processInline = async (seq: number, frame: Buffer): Promise<void> => {
          inFlight++;
          try {
            const result = await deblurAnyImage(
              { data: new Uint8Array(frame), width, height, channels: 3 },
              { ...lockedOptions, watermarkFrameIndex: seq },
            );
            const u8 = floatPlaneToUint8Clamped(result.recovered);
            pending.set(seq, { recovered: toOwnedArrayBuffer(u8), selectedSigma: result.selectedSigma });
          } finally {
            inFlight--;
          }
          await drainPending();
        };

        const dispatchLoop = async (): Promise<void> => {
          if (settled) return;
          while (!readingDone && inFlight < inFlightCap) {
            // Inline mode: serial main-thread processing — no workers, no idle pool.
            if (inlineMode) {
              const frame = await splitter.next();
              if (frame === null) {
                readingDone = true;
                break;
              }
              const seq = nextInSeq++;
              await processInline(seq, frame);
              continue;
            }
            const idle = findIdleWorker();
            if (idle === null) return;
            const frame = await splitter.next();
            if (frame === null) {
              readingDone = true;
              break;
            }
            const seq = nextInSeq++;
            idle.busy = true;
            inFlight++;
            const ab = bufferToTransferableArrayBuffer(frame);
            idle.worker.postMessage({ type: 'process', seq, frame: ab }, [ab]);
          }
          if (readingDone && inFlight === 0) {
            allDispatched = true;
            await drainPending();
          }
        };

        // Kick the dispatcher.
        await dispatchLoop();

        // Wait for everything to drain.
        await allDrained;

        // 10. Shutdown workers, end encoder.stdin, await encoder exit.
        for (const w of workersPool) {
          try { w.worker.postMessage({ type: 'shutdown' }); } catch { /* swallow */ }
        }
        await Promise.all(workersPool.map((w) => w.worker.terminate().catch(() => undefined)));
        workersPool.length = 0;

        encoder.stdin!.end();
        await new Promise<void>((res, rej) => {
          encoder.on('close', (code, signal) => {
            if (code === 0) {
              res();
            } else {
              const stderr = Buffer.concat(encoderStderr).toString('utf8');
              rej(new Error(`ffmpeg encoder exited with code ${code} signal=${signal}: ${stderr}`));
            }
          });
        });

        const elapsedMs = Date.now() - startedAt;
        const elapsedSec = elapsedMs / 1000;
        const averageFps = elapsedSec > 0 ? framesWritten / elapsedSec : 0;

        if (settled) return;
        settled = true;
        resolve({
          framesProcessed: framesWritten,
          selectedSigma: finalSigma,
          width,
          height,
          fps,
          elapsedMs,
          averageFps,
        });
      } catch (err) {
        fail(err instanceof Error ? err : new Error(String(err)));
      }
    })();
  });
}

function toOwnedArrayBuffer(bytes: Uint8Array): ArrayBuffer {
  const out = new ArrayBuffer(bytes.byteLength);
  new Uint8Array(out).set(bytes);
  return out;
}
