// @a0n/gnosis/distributed-inference-host/scripts/write-style-transfer-golden-pngs.ts
/**
 * Writes committed golden PNGs for style-transfer-golden-png tests.
 * Run: pnpm run gnode -- run open-source/gnosis/distributed-inference-host/scripts/write-style-transfer-golden-pngs.ts --export main
 */

import { mkdirSync, writeFileSync } from 'node:fs';
import { join } from 'node:path';
import { rgbFloat01InterleavedToPngBuffer } from '../__tests__/helpers/rgb-float01-to-png';
import { runClassicalStyleTransfer } from '../src/style-transfer-classical';
import { stubTopologyAlphaFromContentRgb } from '../src/style-transfer-topology-glue';

/** Run from `open-source/gnosis/distributed-inference-host` (package root). */
const GOLDEN_DIR = join(process.cwd(), '__tests__', 'golden', 'style-transfer');

function makePlane(w: number, h: number, seed: number): Float32Array {
  const out = new Float32Array(w * h * 3);
  let s = seed >>> 0;
  const rand = () => {
    s = (s * 1664525 + 1013904223) >>> 0;
    return s / 0xffffffff;
  };
  for (let i = 0; i < out.length; i++) out[i] = rand();
  return out;
}

const alphaWaveOff = { enabled: false, edgeThreshold: 0.22, edgeBoost: 0.7, frequency: 2.2, phase: 0 };

/**
 * Runs the gnosis command-line workflow.
 */
export function main(): number {
  mkdirSync(GOLDEN_DIR, { recursive: true });
  const w = 16;
  const h = 16;
  const content = makePlane(w, h, 7);
  const style = makePlane(w, h, 8);

  const classicalNoMesh = runClassicalStyleTransfer({
    content,
    style,
    width: w,
    height: h,
    channels: 3,
    strength: 0.72,
    structurePreservation: 0.4,
    alphaWave: alphaWaveOff,
    sharpen: false,
  });
  writeFileSync(
    join(GOLDEN_DIR, 'classical-no-mesh-16.png'),
    rgbFloat01InterleavedToPngBuffer(classicalNoMesh.stylized, w, h),
  );

  const stub = stubTopologyAlphaFromContentRgb({
    data: content,
    width: w,
    height: h,
    channels: 3,
  });
  const classicalMeshStub = runClassicalStyleTransfer({
    content,
    style,
    width: w,
    height: h,
    channels: 3,
    strength: 0.72,
    structurePreservation: 0.4,
    alphaWave: alphaWaveOff,
    sharpen: false,
    meshGate: {
      topologyAlpha: stub.alpha,
      topologySource: 'stub',
      topologyProvenanceFragment: stub.provenanceFragment,
      depthProvenanceFragment: { note: 'writer-script' },
    },
  });
  writeFileSync(
    join(GOLDEN_DIR, 'classical-mesh-stub-16.png'),
    rgbFloat01InterleavedToPngBuffer(classicalMeshStub.stylized, w, h),
  );

  const n = w * h;
  const depth = new Float32Array(n);
  depth.fill(0.55);
  const classicalMeshDepth = runClassicalStyleTransfer({
    content,
    style,
    width: w,
    height: h,
    channels: 3,
    strength: 0.72,
    structurePreservation: 0.4,
    alphaWave: alphaWaveOff,
    sharpen: false,
    meshGate: {
      topologyAlpha: stub.alpha,
      depthConfidence: depth,
      topologySource: 'stub',
      topologyProvenanceFragment: stub.provenanceFragment,
      depthProvenanceFragment: { mean_confidence: 0.55 },
    },
  });
  writeFileSync(
    join(GOLDEN_DIR, 'classical-mesh-depth-16.png'),
    rgbFloat01InterleavedToPngBuffer(classicalMeshDepth.stylized, w, h),
  );
  return 0;
}
