// @a0n/gnosis/distributed-inference-host/__tests__/helpers/rgb-float01-to-png.ts
/**
 * Deterministic RGB8 PNG (filter 0, zlib) from linear Float32 [0,1] interleaved RGB.
 * Matches `encodeStyledAsPNG` quantization (`round(clamp01(v)*255))`) so golden bytes
 * stay stable without depending on sharp in this helper.
 */

import { crc32, deflateSync } from 'node:zlib';

function writeU32be(n: number): Buffer {
  const b = Buffer.alloc(4);
  b.writeUInt32BE(n >>> 0, 0);
  return b;
}

function pngChunk(type: string, data: Buffer): Buffer {
  const typeBuf = Buffer.from(type, 'binary');
  const crcInput = Buffer.concat([typeBuf, data]);
  const c = (crc32(crcInput) >>> 0) as number;
  return Buffer.concat([writeU32be(data.length), typeBuf, data, writeU32be(c)]);
}

function clamp01(v: number): number {
  if (v <= 0) return 0;
  if (v >= 1) return 1;
  return v;
}

function quantizeByte(v: number): number {
  const c = clamp01(v);
  const x = Math.round(c * 255);
  return x <= 0 ? 0 : x >= 255 ? 255 : x;
}

/** Row-major RGB interleaved floats, length width*height*3. */
export function rgbFloat01InterleavedToPngBuffer(
  rgb: Float32Array,
  width: number,
  height: number,
): Buffer {
  const bpp = 3;
  const expected = width * height * bpp;
  if (rgb.length !== expected) {
    throw new RangeError(`rgbFloat01InterleavedToPngBuffer: expected length ${expected}, got ${rgb.length}`);
  }
  const rowBytes = 1 + width * bpp;
  const raw = Buffer.alloc(rowBytes * height);
  let o = 0;
  for (let y = 0; y < height; y++) {
    raw[o] = 0;
    o++;
    for (let x = 0; x < width; x++) {
      const i = (y * width + x) * bpp;
      raw[o++] = quantizeByte(rgb[i]!);
      raw[o++] = quantizeByte(rgb[i + 1]!);
      raw[o++] = quantizeByte(rgb[i + 2]!);
    }
  }
  const idat = deflateSync(raw, { level: 9 });
  const ihdr = Buffer.alloc(13);
  ihdr.writeUInt32BE(width, 0);
  ihdr.writeUInt32BE(height, 4);
  ihdr.writeUInt8(8, 8);
  ihdr.writeUInt8(2, 9);
  ihdr.writeUInt8(0, 10);
  ihdr.writeUInt8(0, 11);
  ihdr.writeUInt8(0, 12);
  const sig = Buffer.from([0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]);
  return Buffer.concat([
    sig,
    pngChunk('IHDR', ihdr),
    pngChunk('IDAT', idat),
    pngChunk('IEND', Buffer.alloc(0)),
  ]);
}
