// @a0n/gnosis/rtlsdr-mock-sim/distance-ladder.ts
/**
 * The distance ladder: near-Earth landmarks -> Moon -> planets -> deep-space
 * landmarks -> the interstellar wall. Anything that closes at a far rung closes
 * at every nearer rung (monotonicity), so we prove the hard rung and relay the
 * proof down. We also find the break rung (first margin < 0 at a fixed reference
 * config) and report it honestly instead of pretending the far rungs close.
 *
 * Two monotone facts this module asserts numerically (mirrored in Lean):
 *   - link margin is non-increasing in distance (FSPL grows as d^2),
 *   - teleport saving (avoided round-trip light-time) is non-decreasing in distance.
 */
import {
  lightTimeSecFromKm,
  roundTripSecFromKm,
  MARS_RANGE_KM,
} from './mars-light-time';
import {
  BANDS,
  linkBudget,
  type BandName,
  type FecProfile,
  type Modulation,
} from './link-budget';

export type RungClass =
  | 'near-earth'
  | 'moon'
  | 'planet'
  | 'deep-space'
  | 'interstellar';

export interface LadderRung {
  readonly name: string;
  readonly class: RungClass;
  /** Representative Earth<->target range (km). */
  readonly rangeKm: number;
  readonly note?: string;
}

const LIGHT_YEAR_KM = 9_460_730_472_580.8;

/**
 * Rungs in increasing distance. Planet ranges use a representative regime so the
 * ladder is monotone; Mars carries all three regimes via mars-light-time. Values
 * are standard Earth<->body distances (mean unless the landmark is a fixed point).
 */
export const LADDER: readonly LadderRung[] = [
  { name: 'ISS / LEO', class: 'near-earth', rangeKm: 420, note: 'low Earth orbit' },
  { name: 'GEO', class: 'near-earth', rangeKm: 35_786, note: 'geostationary belt' },
  { name: 'Moon', class: 'moon', rangeKm: 384_400, note: '1.28 light-seconds' },
  { name: 'JWST / Sun-Earth L2', class: 'near-earth', rangeKm: 1_500_000 },
  { name: 'Venus (closest)', class: 'planet', rangeKm: 38_000_000 },
  { name: 'Mars (opposition)', class: 'planet', rangeKm: MARS_RANGE_KM.opposition, note: 'reference rung (near)' },
  { name: 'Mercury (mean)', class: 'planet', rangeKm: 155_000_000 },
  { name: 'Mars (mean)', class: 'planet', rangeKm: MARS_RANGE_KM.mean, note: 'reference rung' },
  { name: 'Venus (farthest)', class: 'planet', rangeKm: 261_000_000 },
  { name: 'Mars (conjunction)', class: 'planet', rangeKm: MARS_RANGE_KM.conjunction, note: 'reference rung (far)' },
  { name: 'Jupiter (mean)', class: 'planet', rangeKm: 778_000_000 },
  { name: 'Saturn (mean)', class: 'planet', rangeKm: 1_430_000_000 },
  { name: 'Uranus (mean)', class: 'planet', rangeKm: 2_880_000_000 },
  { name: 'Neptune (mean)', class: 'planet', rangeKm: 4_500_000_000 },
  { name: 'Pluto (mean)', class: 'deep-space', rangeKm: 5_900_000_000 },
  { name: 'New Horizons', class: 'deep-space', rangeKm: 8_400_000_000 },
  { name: 'Voyager 1', class: 'deep-space', rangeKm: 24_600_000_000, note: '~22.8 light-hours' },
  { name: 'Proxima Centauri', class: 'interstellar', rangeKm: 4.2465 * LIGHT_YEAR_KM, note: 'the wall' },
];

export interface LadderConfig {
  readonly band: BandName;
  readonly txPowerDbm: number;
  readonly txDishM: number;
  readonly rxDishM: number;
  readonly systemNoiseTempK: number;
  readonly dataRateBps: number;
  readonly mod: Modulation;
  readonly fec: FecProfile;
  readonly whipReplicas: number;
}

/** A DSN-class reference: 34 m ground dish, 20 W X-band spacecraft, MFSK8 + RS+conv + 4-way Whip. */
export const REFERENCE_CONFIG: LadderConfig = {
  band: 'x-down',
  txPowerDbm: 43, // 20 W
  txDishM: 4,
  rxDishM: 34,
  systemNoiseTempK: 30,
  dataRateBps: 1000,
  mod: 'mfsk8',
  fec: 'rs+conv',
  whipReplicas: 4,
};

export interface LadderRow {
  readonly name: string;
  readonly class: RungClass;
  readonly rangeKm: number;
  readonly oneWaySec: number;
  readonly roundTripSec: number;
  readonly marginDb: number;
  readonly marginCloses: boolean;
  /** Latency a geodesicLength:0 replay avoids (~ the full round-trip). */
  readonly teleportSavingSec: number;
}

/**
 * Builds the Ladder Row.
 */
export function buildLadderRow(rung: LadderRung, cfg: LadderConfig): LadderRow {
  const distanceM = rung.rangeKm * 1000;
  const lb = linkBudget({
    distanceM,
    freqHz: BANDS[cfg.band],
    txPowerDbm: cfg.txPowerDbm,
    txDishM: cfg.txDishM,
    rxDishM: cfg.rxDishM,
    systemNoiseTempK: cfg.systemNoiseTempK,
    dataRateBps: cfg.dataRateBps,
    mod: cfg.mod,
    fec: cfg.fec,
    whipReplicas: cfg.whipReplicas,
  });
  return {
    name: rung.name,
    class: rung.class,
    rangeKm: rung.rangeKm,
    oneWaySec: lightTimeSecFromKm(rung.rangeKm),
    roundTripSec: roundTripSecFromKm(rung.rangeKm),
    marginDb: lb.marginDb,
    marginCloses: lb.marginCloses,
    teleportSavingSec: roundTripSecFromKm(rung.rangeKm),
  };
}

export const LADDER_SWEEP_VERSION = 'gnosis.mars-ladder.v1';

export interface LadderSweep {
  readonly version: typeof LADDER_SWEEP_VERSION;
  readonly config: LadderConfig;
  readonly rows: readonly LadderRow[];
  /** First rung (by increasing distance) whose margin does not close, or null. */
  readonly breakRung: string | null;
  /** Margins are non-increasing with distance. */
  readonly marginMonotoneNonIncreasing: boolean;
  /** Teleport savings are non-decreasing with distance. */
  readonly savingMonotoneNonDecreasing: boolean;
}

/**
 * Handles the gnosis sweep Ladder workflow.
 */
export function sweepLadder(cfg: LadderConfig = REFERENCE_CONFIG): LadderSweep {
  const sorted = [...LADDER].sort((a, b) => a.rangeKm - b.rangeKm);
  const rows = sorted.map((r) => buildLadderRow(r, cfg));
  let breakRung: string | null = null;
  for (const row of rows) {
    if (!row.marginCloses) {
      breakRung = row.name;
      break;
    }
  }
  let marginMono = true;
  let savingMono = true;
  for (let i = 1; i < rows.length; i++) {
    if (rows[i]!.marginDb > rows[i - 1]!.marginDb + 1e-9) marginMono = false;
    if (rows[i]!.teleportSavingSec < rows[i - 1]!.teleportSavingSec - 1e-9) savingMono = false;
  }
  return {
    version: LADDER_SWEEP_VERSION,
    config: cfg,
    rows,
    breakRung,
    marginMonotoneNonIncreasing: marginMono,
    savingMonotoneNonDecreasing: savingMono,
  };
}
