// @a0n/gnosis/rtlsdr-mock-sim/fusion.ts
import type { Rtl433Event } from './rtl433';

export interface BatchCaptureSummaryRow {
  readonly file: string;
  readonly wavOut?: string;
  readonly captureMtimeMs?: number;
  readonly [k: string]: unknown;
}

export interface CaptureEventMatch {
  readonly eventIndex: number;
  readonly captureIndex: number;
  readonly deltaMs: number;
}

/**
 * Handles the gnosis correlate Events To Captures workflow.
 */
export function correlateEventsToCaptures(
  captures: readonly BatchCaptureSummaryRow[],
  events: readonly Rtl433Event[],
  maxDeltaMs: number,
): CaptureEventMatch[] {
  if (!Number.isFinite(maxDeltaMs) || maxDeltaMs < 0) {
    throw new Error('correlateEventsToCaptures: maxDeltaMs must be >= 0');
  }
  const matches: CaptureEventMatch[] = [];
  for (let ei = 0; ei < events.length; ei++) {
    const ev = events[ei]!;
    if (ev.whenIso === null) continue;
    const t = Date.parse(ev.whenIso);
    if (!Number.isFinite(t)) continue;
    let bestCi = -1;
    let bestDelta = Number.POSITIVE_INFINITY;
    for (let ci = 0; ci < captures.length; ci++) {
      const c = captures[ci]!;
      const mt = c.captureMtimeMs;
      if (typeof mt !== 'number' || !Number.isFinite(mt)) continue;
      const d = Math.abs(mt - t);
      if (d < bestDelta) {
        bestDelta = d;
        bestCi = ci;
      }
    }
    if (bestCi >= 0 && bestDelta <= maxDeltaMs) {
      matches.push({ eventIndex: ei, captureIndex: bestCi, deltaMs: bestDelta });
    }
  }
  return matches;
}

/**
 * Handles the gnosis summarize Capture Match Counts workflow.
 */
export function summarizeCaptureMatchCounts(
  captures: readonly BatchCaptureSummaryRow[],
  matches: readonly CaptureEventMatch[],
): number[] {
  const counts = Array.from({ length: captures.length }, () => 0);
  for (const m of matches) {
    if (m.captureIndex >= 0 && m.captureIndex < counts.length) {
      counts[m.captureIndex]! += 1;
    }
  }
  return counts;
}
