// @a0n/gnosis/rtlsdr-mock-sim/interleaver.ts
/**
 * Block (matrix) interleaver. Write rows, read columns -> a burst of consecutive
 * channel errors is spread across many RS codewords/symbols so the outer code
 * sees only scattered single-symbol errors. The CCSDS concatenation pattern.
 */

/** Interleave bytes through an rows x cols matrix (zero-padded). Returns rows*cols bytes. */
export function blockInterleave(bytes: Uint8Array, rows: number, cols: number): Uint8Array {
  if (rows < 1 || cols < 1) throw new Error('blockInterleave: rows,cols >= 1');
  const size = rows * cols;
  if (bytes.length > size) {
    throw new Error(`blockInterleave: ${bytes.length} bytes exceed ${rows}x${cols}=${size}`);
  }
  const grid = new Uint8Array(size);
  grid.set(bytes); // row-major write
  const out = new Uint8Array(size);
  let o = 0;
  for (let c = 0; c < cols; c++) {
    for (let r = 0; r < rows; r++) {
      out[o++] = grid[r * cols + c]!;
    }
  }
  return out;
}

/** Inverse of blockInterleave; returns the first `originalLen` bytes. */
export function blockDeinterleave(
  bytes: Uint8Array,
  rows: number,
  cols: number,
  originalLen: number,
): Uint8Array {
  const size = rows * cols;
  if (bytes.length !== size) {
    throw new Error(`blockDeinterleave: expected ${size} bytes, got ${bytes.length}`);
  }
  const grid = new Uint8Array(size);
  let i = 0;
  for (let c = 0; c < cols; c++) {
    for (let r = 0; r < rows; r++) {
      grid[r * cols + c] = bytes[i++]!;
    }
  }
  return grid.slice(0, originalLen);
}
