// @a0n/gnosis/rtlsdr-mock-sim/iq.ts
/**
 * rtl_sdr-compatible interleaved uint8 IQ (offset-binary around 128).
 * Matches what Linux `rtl2832` + `rtl_sdr` userspace typically records as raw u8.
 */

export function u8FromComplexSample(re: number, im: number): [number, number] {
  const i = Math.min(255, Math.max(0, Math.round(128 + 127 * re)));
  const q = Math.min(255, Math.max(0, Math.round(128 + 127 * im)));
  return [i, q];
}

/**
 * Handles the gnosis complex From U8 Pair workflow.
 */
export function complexFromU8Pair(iByte: number, qByte: number): { re: number; im: number } {
  return { re: (iByte - 128) / 128, im: (qByte - 128) / 128 };
}

/**
 * Handles the gnosis append Complex U8 workflow.
 */
export function appendComplexU8(
  out: number[],
  re: number,
  im: number,
): void {
  const [i, q] = u8FromComplexSample(re, im);
  out.push(i, q);
}

/**
 * Handles the gnosis u8 Array From Pairs workflow.
 */
export function u8ArrayFromPairs(pairs: readonly number[]): Uint8Array {
  if (pairs.length % 2 !== 0) {
    throw new Error('u8ArrayFromPairs: expected even length [I,Q,...]');
  }
  return Uint8Array.from(pairs);
}
