// @a0n/gnosis/rtlsdr-mock-sim/mars-channel.ts
/**
 * Mars-channel emulation. Adds complex AWGN at a fixed per-component RMS (the SDR
 * ADC noise fill) to interleaved offset-binary u8 IQ, optionally a constant
 * Doppler spin. The transmitted signal amplitude (set by the modem) encodes the
 * target Eb/N0 against this noise floor -- see link-budget.signalAmplitudeForEbN0.
 * Deterministic when seeded so the raw-vs-FEC A/B is reproducible in CI. The
 * teleport path never runs through here; this only degrades the bench link.
 */
import { complexFromU8Pair, u8FromComplexSample } from './iq';
import { gaussianPairFrom, mulberry32, type UniformRng } from './noise';

export interface MarsChannelOptions {
  readonly noiseRms: number;
  readonly sampleRateHz?: number;
  readonly dopplerHz?: number;
  readonly seed?: number;
}

export interface MarsChannelResult {
  readonly iq: Uint8Array;
  readonly noiseRms: number;
  readonly clipFraction: number;
}

/** Apply AWGN (+ optional Doppler) to a u8 IQ buffer. */
export function applyMarsChannel(iqU8: Uint8Array, opts: MarsChannelOptions): MarsChannelResult {
  if (iqU8.length % 2 !== 0) throw new Error('applyMarsChannel: odd IQ length');
  const rng: UniformRng = opts.seed === undefined ? Math.random : mulberry32(opts.seed);
  const out = new Uint8Array(iqU8.length);
  const wDoppler =
    opts.dopplerHz && opts.sampleRateHz
      ? (Math.PI * 2 * opts.dopplerHz) / opts.sampleRateHz
      : 0;
  let clipped = 0;
  for (let i = 0, p = 0; p + 1 < iqU8.length; i++, p += 2) {
    const { re, im } = complexFromU8Pair(iqU8[p]!, iqU8[p + 1]!);
    let sre = re;
    let sim = im;
    if (wDoppler !== 0) {
      const c = Math.cos(wDoppler * i);
      const s = Math.sin(wDoppler * i);
      sre = re * c - im * s;
      sim = re * s + im * c;
    }
    const [g0, g1] = gaussianPairFrom(rng);
    const nre = sre + opts.noiseRms * g0;
    const nim = sim + opts.noiseRms * g1;
    if (Math.abs(nre) > 1 || Math.abs(nim) > 1) clipped++;
    const [bi, bq] = u8FromComplexSample(nre, nim);
    out[p] = bi;
    out[p + 1] = bq;
  }
  return { iq: out, noiseRms: opts.noiseRms, clipFraction: clipped / (iqU8.length / 2) };
}
