// @a0n/gnosis/rtlsdr-mock-sim/mars-cold-roundtrip.ts
/**
 * Cold round-trip model: the latency a request pays when it must traverse the
 * interplanetary link instead of teleporting a cache key. This is the baseline
 * the geodesicLength:0 replay avoids. We never wait real minutes -- the modeled
 * one-way delay is scaled (delayScale) so a demo runs in subseconds while the
 * certificate records the true modeled seconds.
 *
 * The transport is delay-tolerant store-carry-forward (the @a0n/aeon DTN stack:
 * spray-and-wait bundles, TTL, magic-number routing). Here we model the delay it
 * incurs; the DTN bundle framing lives in open-source/aeon/src/dtn.
 */
import { roundTripSecFromKm, lightTimeSecFromKm } from './mars-light-time';

export interface ColdRoundTrip {
  readonly rangeKm: number;
  readonly oneWaySec: number;
  readonly roundTripSec: number;
  /** Wall-clock the demo actually waited (= roundTripSec * delayScale). */
  readonly modeledWallClockSec: number;
  readonly delayScale: number;
  /** DTN bundle TTL recommended for this hop (2x one-way + slack). */
  readonly recommendedTtlSec: number;
}

/**
 * Handles the gnosis model Cold Round Trip workflow.
 */
export function modelColdRoundTrip(rangeKm: number, delayScale = 1): ColdRoundTrip {
  const oneWay = lightTimeSecFromKm(rangeKm);
  const rt = roundTripSecFromKm(rangeKm);
  return {
    rangeKm,
    oneWaySec: oneWay,
    roundTripSec: rt,
    modeledWallClockSec: rt * delayScale,
    delayScale,
    recommendedTtlSec: Math.ceil(rt + 60),
  };
}
