// @a0n/gnosis/rtlsdr-mock-sim/mfsk-modem.ts
/**
 * Non-coherent M-FSK modem for the cache-key frame. Each symbol is one of M
 * orthogonal tones (spacing = baud), detected by a per-symbol single-bin DFT
 * argmax -- the proven 66 kHz tone detector, repeated per symbol. No carrier
 * phase recovery; the NESDR 0.5 ppm TCXO leaves residual CFO << tone spacing.
 *
 * Frame: [SYNC syms][HEADER: uint16 blockLen, triplicated][BODY: FEC(payload+crc16)].
 * BODY length is derived deterministically from blockLen + FEC profile.
 */
import { complexFromU8Pair } from './iq';
import {
  codedBitLength,
  decodeFec,
  encodeFec,
  type FecProfile,
} from './fec-codec';

export type MfskMod = 'mfsk8' | 'mfsk4';

export interface MfskOptions {
  readonly mod: MfskMod;
  readonly iqRateHz: number;
  readonly baud: number;
  readonly baseFreqHz: number;
  readonly fec: FecProfile;
  /** Tone amplitude in [-1,1] u8 scale. Encodes Eb/N0 vs the channel noise fill. */
  readonly amplitude?: number;
}

export const DEFAULT_MFSK: Omit<MfskOptions, 'fec'> = {
  mod: 'mfsk8',
  iqRateHz: 2_400_000, // valid for both Pluto TX (min ~2.083 MS/s) and NESDR RX
  baud: 8000, // sps = 300 (integer)
  baseFreqHz: 96_000, // multiple of baud -> orthogonal, integer cycles/symbol
  amplitude: 0.6,
};

// Distinctive fixed sync pattern (values reduced mod M at use).
const SYNC_BASE = [0, 7, 1, 6, 2, 5, 3, 4, 7, 0, 6, 1, 5, 2, 4, 3, 0, 5, 2, 7, 1, 4, 3, 6];
const HEADER_REPEAT = 3;
const HEADER_BITS = 16;

/**
 * Handles the gnosis m For Mod workflow.
 */
export function mForMod(mod: MfskMod): number {
  return mod === 'mfsk8' ? 8 : 4;
}
/**
 * Handles the gnosis bps For Mod workflow.
 */
export function bpsForMod(mod: MfskMod): number {
  return mod === 'mfsk8' ? 3 : 2;
}
function spsForOpts(o: { iqRateHz: number; baud: number }): number {
  const sps = Math.round(o.iqRateHz / o.baud);
  if (sps < 8) throw new Error(`mfsk: samplesPerSymbol ${sps} too small`);
  return sps;
}

/** CRC-16/CCITT-FALSE over bytes (poly 0x1021, init 0xFFFF). */
export function crc16Bytes(bytes: Uint8Array): number {
  let crc = 0xffff;
  for (let i = 0; i < bytes.length; i++) {
    crc ^= bytes[i]! << 8;
    for (let b = 0; b < 8; b++) {
      crc = (crc & 0x8000) !== 0 ? ((crc << 1) ^ 0x1021) & 0xffff : (crc << 1) & 0xffff;
    }
  }
  return crc & 0xffff;
}

function bitsToSymbols(bits: Uint8Array, bps: number): Uint8Array {
  const nSym = Math.ceil(bits.length / bps);
  const out = new Uint8Array(nSym);
  for (let s = 0; s < nSym; s++) {
    let v = 0;
    for (let b = 0; b < bps; b++) {
      const idx = s * bps + b;
      v = (v << 1) | (idx < bits.length ? bits[idx]! & 1 : 0);
    }
    out[s] = v;
  }
  return out;
}
function symbolsToBits(syms: Uint8Array, bps: number, nBits: number): Uint8Array {
  const out = new Uint8Array(nBits);
  let o = 0;
  for (let s = 0; s < syms.length && o < nBits; s++) {
    for (let b = bps - 1; b >= 0 && o < nBits; b--) out[o++] = (syms[s]! >> b) & 1;
  }
  return out;
}

function headerSymbols(blockLen: number, bps: number): Uint8Array {
  const bits = new Uint8Array(HEADER_BITS * HEADER_REPEAT);
  for (let r = 0; r < HEADER_REPEAT; r++) {
    for (let b = 0; b < HEADER_BITS; b++) {
      bits[r * HEADER_BITS + b] = (blockLen >> (HEADER_BITS - 1 - b)) & 1;
    }
  }
  return bitsToSymbols(bits, bps);
}
function headerSymCount(bps: number): number {
  return Math.ceil((HEADER_BITS * HEADER_REPEAT) / bps);
}

function toneHz(base: number, spacing: number, s: number): number {
  return base + s * spacing;
}

/** Total IQ samples (pairs) a frame occupies for a given payload length. */
export function frameSampleCount(payloadLen: number, o: MfskOptions): number {
  const bps = bpsForMod(o.mod);
  const sps = spsForOpts(o);
  const blockLen = payloadLen + 2; // + crc16
  const bodySyms = Math.ceil(codedBitLength(blockLen, o.fec) / bps);
  const totalSyms = SYNC_BASE.length + headerSymCount(bps) + bodySyms;
  return totalSyms * sps;
}

/** Append `sps` samples of a unit tone for symbol `s` to a u8 IQ array, phase-continuous. */
function synthSymbol(
  out: number[],
  s: number,
  sps: number,
  base: number,
  spacing: number,
  fs: number,
  phase0: number,
  amp: number,
): number {
  const w = (Math.PI * 2 * toneHz(base, spacing, s)) / fs;
  let ph = phase0;
  for (let n = 0; n < sps; n++) {
    const re = amp * Math.cos(ph);
    const im = amp * Math.sin(ph);
    out.push(
      Math.min(255, Math.max(0, Math.round(128 + 127 * re))),
      Math.min(255, Math.max(0, Math.round(128 + 127 * im))),
    );
    ph += w;
  }
  return ph;
}

/**
 * Handles the gnosis modulate Frame workflow.
 */
export function modulateFrame(payload: Uint8Array, o: MfskOptions): Uint8Array {
  const M = mForMod(o.mod);
  const bps = bpsForMod(o.mod);
  const sps = spsForOpts(o);
  const spacing = o.baud;
  // body = payload + crc16
  const crc = crc16Bytes(payload);
  const body = new Uint8Array(payload.length + 2);
  body.set(payload);
  body[payload.length] = (crc >> 8) & 0xff;
  body[payload.length + 1] = crc & 0xff;
  const blockLen = body.length;
  const coded = encodeFec(body, o.fec);
  const bodySyms = bitsToSymbols(coded, bps);
  const sync = Uint8Array.from(SYNC_BASE, (v) => v % M);
  const hdr = headerSymbols(blockLen, bps);
  const all = new Uint8Array(sync.length + hdr.length + bodySyms.length);
  all.set(sync, 0);
  all.set(hdr, sync.length);
  all.set(bodySyms, sync.length + hdr.length);
  const amp = o.amplitude ?? 0.6;
  const out: number[] = [];
  let phase = 0;
  for (let i = 0; i < all.length; i++) {
    phase = synthSymbol(out, all[i]!, sps, o.baseFreqHz, spacing, o.iqRateHz, phase, amp);
  }
  return Uint8Array.from(out);
}

/** Single-bin DFT power at tone `s` over a window [start, start+sps). */
function tonePower(
  re: Float64Array,
  im: Float64Array,
  start: number,
  sps: number,
  s: number,
  base: number,
  spacing: number,
  fs: number,
): number {
  const w = (Math.PI * 2 * toneHz(base, spacing, s)) / fs;
  let cr = 0;
  let ci = 0;
  for (let n = 0; n < sps; n++) {
    const c = Math.cos(w * n);
    const sn = Math.sin(w * n);
    const xr = re[start + n]!;
    const xi = im[start + n]!;
    // multiply by conj(tone) = (c, -sn)
    cr += xr * c + xi * sn;
    ci += xi * c - xr * sn;
  }
  return cr * cr + ci * ci;
}

function detectSymbol(
  re: Float64Array,
  im: Float64Array,
  start: number,
  sps: number,
  M: number,
  base: number,
  spacing: number,
  fs: number,
): { sym: number; conf: number } {
  let best = -1;
  let bestP = -1;
  let total = 0;
  for (let s = 0; s < M; s++) {
    const p = tonePower(re, im, start, sps, s, base, spacing, fs);
    total += p;
    if (p > bestP) {
      bestP = p;
      best = s;
    }
  }
  return { sym: best, conf: total > 0 ? bestP / total : 0 };
}

function syncScore(
  re: Float64Array,
  im: Float64Array,
  offset: number,
  sps: number,
  sync: Uint8Array,
  M: number,
  base: number,
  spacing: number,
  fs: number,
): number {
  let score = 0;
  for (let k = 0; k < sync.length; k++) {
    const start = offset + k * sps;
    if (start + sps > re.length) return -1;
    let total = 0;
    const exp = tonePower(re, im, start, sps, sync[k]!, base, spacing, fs);
    for (let s = 0; s < M; s++) total += tonePower(re, im, start, sps, s, base, spacing, fs);
    score += total > 0 ? exp / total : 0;
  }
  return score / sync.length;
}

export interface DemodResult {
  readonly ok: boolean;
  readonly payload: Uint8Array | null;
  readonly blockLen: number;
  readonly syncOffset: number;
  readonly syncScore: number;
  readonly rsErrors: number;
  readonly reason?: string;
}

/**
 * Handles the gnosis demodulate Frame workflow.
 */
export function demodulateFrame(
  iqU8: Uint8Array,
  o: MfskOptions,
  searchFromSample = 0,
  maxSearchSamples = Number.POSITIVE_INFINITY,
): DemodResult {
  const M = mForMod(o.mod);
  const bps = bpsForMod(o.mod);
  const sps = spsForOpts(o);
  const spacing = o.baud;
  const nPairs = Math.floor(iqU8.length / 2);
  const pad = sps; // zero-pad tail so a few-sample sync offset never truncates
  const re = new Float64Array(nPairs + pad);
  const im = new Float64Array(nPairs + pad);
  for (let p = 0; p < nPairs; p++) {
    const c = complexFromU8Pair(iqU8[2 * p]!, iqU8[2 * p + 1]!);
    re[p] = c.re;
    im[p] = c.im;
  }
  const sync = Uint8Array.from(SYNC_BASE, (v) => v % M);
  const hdrCount = headerSymCount(bps);

  // coarse + fine sync search (bounded; early-accept a strong peak)
  const frameMinSyms = sync.length + hdrCount + 1;
  const maxStart = Math.min(
    Math.max(0, nPairs - frameMinSyms * sps),
    searchFromSample + maxSearchSamples,
  );
  let bestOff = -1;
  let bestScore = -1;
  const coarse = Math.max(8, sps >> 2);
  const EARLY_ACCEPT = 0.55;
  for (let off = searchFromSample; off <= maxStart; off += coarse) {
    const sc = syncScore(re, im, off, sps, sync, M, o.baseFreqHz, spacing, o.iqRateHz);
    if (sc > bestScore) {
      bestScore = sc;
      bestOff = off;
    }
    if (sc >= EARLY_ACCEPT) break; // strong peak found; refine around it
  }
  if (bestOff < 0) {
    return { ok: false, payload: null, blockLen: 0, syncOffset: -1, syncScore: bestScore, rsErrors: 0, reason: 'no sync' };
  }
  for (let off = Math.max(0, bestOff - coarse); off <= bestOff + coarse; off++) {
    if (off > maxStart) break;
    const sc = syncScore(re, im, off, sps, sync, M, o.baseFreqHz, spacing, o.iqRateHz);
    if (sc > bestScore) {
      bestScore = sc;
      bestOff = off;
    }
  }

  // read header (3 copies, majority on the decoded length)
  const hdrStart = bestOff + sync.length * sps;
  const hdrSyms = new Uint8Array(hdrCount);
  for (let k = 0; k < hdrCount; k++) {
    const d = detectSymbol(re, im, hdrStart + k * sps, sps, M, o.baseFreqHz, spacing, o.iqRateHz);
    hdrSyms[k] = d.sym;
  }
  const hdrBits = symbolsToBits(hdrSyms, bps, HEADER_BITS * HEADER_REPEAT);
  const lens: number[] = [];
  for (let r = 0; r < HEADER_REPEAT; r++) {
    let v = 0;
    for (let b = 0; b < HEADER_BITS; b++) v = (v << 1) | hdrBits[r * HEADER_BITS + b]!;
    lens.push(v);
  }
  const blockLen = majority(lens);
  if (blockLen <= 2 || blockLen > 255) {
    return { ok: false, payload: null, blockLen, syncOffset: bestOff, syncScore: bestScore, rsErrors: 0, reason: `bad header len ${blockLen}` };
  }

  const codedBits = codedBitLength(blockLen, o.fec);
  const bodySymCount = Math.ceil(codedBits / bps);
  const bodyStart = hdrStart + hdrCount * sps;
  if (bodyStart + bodySymCount * sps > nPairs + pad) {
    return { ok: false, payload: null, blockLen, syncOffset: bestOff, syncScore: bestScore, rsErrors: 0, reason: 'frame truncated' };
  }
  const bodySyms = new Uint8Array(bodySymCount);
  for (let k = 0; k < bodySymCount; k++) {
    const d = detectSymbol(re, im, bodyStart + k * sps, sps, M, o.baseFreqHz, spacing, o.iqRateHz);
    bodySyms[k] = d.sym;
  }
  const codedStream = symbolsToBits(bodySyms, bps, codedBits);
  let body: Uint8Array;
  let rsErrors = 0;
  try {
    const dec = decodeFec(codedStream, o.fec, blockLen);
    body = dec.block;
    rsErrors = dec.rsErrors;
  } catch (e) {
    return { ok: false, payload: null, blockLen, syncOffset: bestOff, syncScore: bestScore, rsErrors: 0, reason: `fec: ${(e as Error).message}` };
  }
  const payload = body.subarray(0, body.length - 2);
  const gotCrc = ((body[body.length - 2]! << 8) | body[body.length - 1]!) & 0xffff;
  const wantCrc = crc16Bytes(payload);
  if (gotCrc !== wantCrc) {
    return { ok: false, payload: null, blockLen, syncOffset: bestOff, syncScore: bestScore, rsErrors, reason: 'crc mismatch' };
  }
  return { ok: true, payload: payload.slice(), blockLen, syncOffset: bestOff, syncScore: bestScore, rsErrors };
}

function majority(vals: number[]): number {
  const counts = new Map<number, number>();
  let best = vals[0]!;
  let bestC = 0;
  for (const v of vals) {
    const c = (counts.get(v) ?? 0) + 1;
    counts.set(v, c);
    if (c > bestC) {
      bestC = c;
      best = v;
    }
  }
  return best;
}
