// @a0n/gnosis/rtlsdr-mock-sim/pipeline.ts
import { readFileSync, writeFileSync } from 'node:fs';
import {
  demodulateFmDiscriminatorDecimated,
  estimateMeanInstFreqHz,
  pcm16MonoFromFloatAudio,
  scoreToneProjection,
  scoreVoiceBandEnergy,
} from './fm-demod';
import { generateSyntheticFmNwrLikeIqU8 } from './synthetic';
import { writeWavPcm16MonoLe } from './wav';

export interface ForkRaceFoldIfOptions {
  readonly iqSampleRateHz: number;
  readonly audioSampleRateHz: number;
  /** Hz offsets applied at complex baseband before demod (multi-look IF search). */
  readonly ifOffsetsHz: readonly number[];
  /** When set, branch score uses coherent tone projection (synthetic FM lab). */
  readonly scoreMessageHz?: number;
}

export interface ForkRaceFoldIfResult {
  readonly bestOffsetHz: number;
  readonly bestScore: number;
  readonly pcm16: Int16Array;
  /** Mean IQ spin estimate (Hz); tight on CW, loose hint under FM. */
  readonly iqSpinHintHz: number;
}

/**
 * Fork: rotate baseband by several IF hypotheses.
 * Race: score decimated FM audio "voice-band" energy.
 * Fold: pick winner, emit PCM16 mono.
 */
export function forkRaceFoldIfDemod(
  iqU8: Uint8Array,
  opts: ForkRaceFoldIfOptions,
): ForkRaceFoldIfResult {
  let best: ForkRaceFoldIfResult | undefined;
  const twoPi = Math.PI * 2;
  const iqSpinHintHz = estimateMeanInstFreqHz(
    iqU8,
    opts.iqSampleRateHz,
    Math.min(300_000, Math.floor(iqU8.length / 2) - 1),
  );
  for (const offHz of opts.ifOffsetsHz) {
    const rot = rotateIqFrequency(iqU8, opts.iqSampleRateHz, offHz, twoPi);
    const audio = demodulateFmDiscriminatorDecimated(rot, {
      iqSampleRateHz: opts.iqSampleRateHz,
      audioSampleRateHz: opts.audioSampleRateHz,
    });
    const score =
      opts.scoreMessageHz !== undefined
        ? scoreToneProjection(audio, opts.scoreMessageHz, opts.audioSampleRateHz)
        : scoreVoiceBandEnergy(audio);
    const pcm16 = pcm16MonoFromFloatAudio(audio, 12);
    if (best === undefined || score > best.bestScore) {
      best = { bestOffsetHz: offHz, bestScore: score, pcm16, iqSpinHintHz };
    }
  }
  if (best === undefined) {
    throw new Error('forkRaceFoldIfDemod: no IF branches');
  }
  return { ...best, iqSpinHintHz };
}

/**
 * Handles the gnosis rotate Iq Frequency workflow.
 */
export function rotateIqFrequency(
  iqU8: Uint8Array,
  fs: number,
  deltaHz: number,
  twoPi: number = Math.PI * 2,
): Uint8Array {
  if (deltaHz === 0) return iqU8;
  const out = new Uint8Array(iqU8.length);
  const w = (twoPi * deltaHz) / fs;
  for (let i = 0, p = 0; p + 1 < iqU8.length; i++, p += 2) {
    const re = (iqU8[p]! - 128) / 128;
    const im = (iqU8[p + 1]! - 128) / 128;
    const c = Math.cos(w * i);
    const s = Math.sin(w * i);
    const re2 = re * c + im * s;
    const im2 = im * c - re * s;
    const ir = Math.min(255, Math.max(0, Math.round(128 + 127 * re2)));
    const qr = Math.min(255, Math.max(0, Math.round(128 + 127 * im2)));
    out[p] = ir;
    out[p + 1] = qr;
  }
  return out;
}

/**
 * Loads the Rtl U8 Capture.
 */
export function loadRtlU8Capture(path: string): Uint8Array {
  return readFileSync(path);
}

/**
 * Handles the gnosis write Rtl U8 Capture workflow.
 */
export function writeRtlU8Capture(path: string, data: Uint8Array): void {
  writeFileSync(path, data);
}

/**
 * Handles the gnosis write Wav File workflow.
 */
export function writeWavFile(path: string, pcm: Int16Array, sampleRateHz: number): void {
  writeFileSync(path, writeWavPcm16MonoLe(pcm, sampleRateHz));
}

/**
 * Handles the gnosis run Synthetic Pipeline workflow.
 */
export function runSyntheticPipeline(args: {
  readonly seconds: number;
  readonly iqRate: number;
  readonly audioRate: number;
  readonly noiseRms: number;
  readonly ifErrorHz: number;
  readonly messageToneHz: number;
  readonly u8Out: string;
  readonly wavOut: string;
}): ForkRaceFoldIfResult {
  const iq = generateSyntheticFmNwrLikeIqU8({
    sampleRateHz: args.iqRate,
    durationSec: args.seconds,
    noiseRms: args.noiseRms,
    ifErrorHz: args.ifErrorHz,
    fmDeviationHz: 7500,
    messageToneHz: args.messageToneHz,
  });
  writeRtlU8Capture(args.u8Out, iq);
  const hint = estimateMeanInstFreqHz(
    iq,
    args.iqRate,
    Math.min(300_000, Math.floor(iq.length / 2) - 1),
  );
  const span = 6000;
  const step = 500;
  const offsets: number[] = [];
  for (let o = -span; o <= span; o += step) {
    offsets.push(hint + o);
  }
  const fold = forkRaceFoldIfDemod(iq, {
    iqSampleRateHz: args.iqRate,
    audioSampleRateHz: args.audioRate,
    ifOffsetsHz: offsets,
    scoreMessageHz: args.messageToneHz,
  });
  writeWavFile(args.wavOut, fold.pcm16, args.audioRate);
  return fold;
}
