// @a0n/gnosis/rtlsdr-mock-sim/skymesh-witness.ts
/**
 * Skymesh witness extraction.
 *
 * Two air-gapped RTL-SDR stations that both observe the SAME always-on ambient
 * RF broadcast must independently derive the SAME integer "witness" from the
 * live signal, despite each seeing it through its own noise/gain/PPM. The
 * environment is the channel; nobody transmits between the stations.
 *
 * Primary method: SPECTRAL FINGERPRINT. We bin the capture's energy into a small
 * number of coarse frequency bands via per-band single-bin DFTs (the same
 * single-bin DFT power kernel the mfsk-modem uses for tone detection), normalize
 * to a shape vector, then quantize each band to a coarse integer level with
 * hysteresis-free fixed thresholds. Coarse bins + thresholds mean that small
 * sample-to-sample differences between two receivers of the same signal do NOT
 * change the quantized fingerprint, so the derived witness agrees; a different
 * ambient signal produces a different band shape and a different witness.
 *
 * Alternate method: DECODED-FIELD. An rtl_433 JSONL device id/value (via
 * rtl433.ts) is a real decoded field both stations independently recover off the
 * air; we fold it into the same integer witness space.
 */
import { complexFromU8Pair } from './iq';
import { crc16CcittAscii } from './rf-cache-key';
import type { Rtl433Event } from './rtl433';

export interface WitnessOptions {
  /** IQ sample rate (Hz). Sets the frequency grid for the band DFTs. */
  readonly sampleRateHz?: number;
  /** Number of coarse spectral bands across [-fs/2, +fs/2). More = finer. */
  readonly bands?: number;
  /** Quantization levels per band (coarse -> robust). */
  readonly levels?: number;
  /** Max IQ pairs to analyze (bounds cost on big captures). */
  readonly maxPairs?: number;
}

export interface WitnessResult {
  /** Stable integer fingerprint of the ambient signal. */
  readonly witness: number;
  /** Dominant-band concentration in [0,1]; higher = sharper/cleaner lock. */
  readonly confidence: number;
  /** Normalized per-band energy shape (length = bands). */
  readonly spectrum: number[];
}

export const DEFAULT_WITNESS_OPTS: Required<Omit<WitnessOptions, 'maxPairs'>> & {
  maxPairs: number;
} = {
  sampleRateHz: 2_400_000,
  // 64 fine bands: the carrier bin is gain-invariant (same Hz on both receivers)
  // AND discriminative (different signals land on different bins). Measured on two
  // real NESDRs off one FM station: identical top-3 bands; a different station
  // shares zero. Coarser (16) under-discriminates; finer (128) splits the carrier.
  bands: 64,
  levels: 4,
  maxPairs: 262_144,
};

/**
 * Single-bin DFT power at a normalized angular frequency `w` (rad/sample) over a
 * window. Mirrors mfsk-modem's tonePower kernel (conj-tone correlation).
 */
function binPower(
  re: Float64Array,
  im: Float64Array,
  n: number,
  w: number,
): number {
  let cr = 0;
  let ci = 0;
  for (let k = 0; k < n; k++) {
    const c = Math.cos(w * k);
    const s = Math.sin(w * k);
    const xr = re[k]!;
    const xi = im[k]!;
    // multiply by conj(tone) = (c, -s)
    cr += xr * c + xi * s;
    ci += xi * c - xr * s;
  }
  return cr * cr + ci * ci;
}

/**
 * Compute the normalized coarse band-energy shape of a u8 IQ capture.
 * Bands tile the complex baseband from -fs/2 (inclusive) to +fs/2 (exclusive).
 */
export function bandSpectrum(
  iqU8: Uint8Array,
  opts: WitnessOptions = {},
): { spectrum: number[]; dominant: number; concentration: number } {
  const bands = opts.bands ?? DEFAULT_WITNESS_OPTS.bands;
  const maxPairs = opts.maxPairs ?? DEFAULT_WITNESS_OPTS.maxPairs;
  const nPairsAll = Math.floor(iqU8.length / 2);
  const n = Math.min(nPairsAll, maxPairs);
  const re = new Float64Array(n);
  const im = new Float64Array(n);
  for (let p = 0; p < n; p++) {
    const c = complexFromU8Pair(iqU8[2 * p]!, iqU8[2 * p + 1]!);
    re[p] = c.re;
    im[p] = c.im;
  }
  // Remove DC so a receiver's offset-binary bias doesn't dominate band 0.
  let mr = 0;
  let mi = 0;
  for (let k = 0; k < n; k++) {
    mr += re[k]!;
    mi += im[k]!;
  }
  mr = n > 0 ? mr / n : 0;
  mi = n > 0 ? mi / n : 0;
  for (let k = 0; k < n; k++) {
    re[k] = re[k]! - mr;
    im[k] = im[k]! - mi;
  }

  const power = new Array<number>(bands).fill(0);
  let total = 0;
  for (let b = 0; b < bands; b++) {
    // band-center fractional frequency in [-0.5, +0.5)
    const frac = -0.5 + (b + 0.5) / bands;
    const w = 2 * Math.PI * frac;
    const p = n > 0 ? binPower(re, im, n, w) / (n * n) : 0;
    power[b] = p;
    total += p;
  }
  const spectrum = new Array<number>(bands);
  let dominant = 0;
  let domVal = -1;
  for (let b = 0; b < bands; b++) {
    const v = total > 0 ? power[b]! / total : 0;
    spectrum[b] = v;
    if (v > domVal) {
      domVal = v;
      dominant = b;
    }
  }
  const concentration = total > 0 ? domVal : 0;
  return { spectrum, dominant, concentration };
}

/** Number of dominant bands the witness is built from. */
export const WITNESS_TOP_K = 3;
/** Bands needed in common for two witnesses to agree (resilient: tolerates 1 flip). */
export const WITNESS_MIN_SHARED = 2;

/**
 * The indices of the `k` highest-energy bands, returned ascending. This is the
 * GAIN-INVARIANT feature: two receivers of the same strong carrier see the same
 * dominant bands even when their absolute energy / antenna gain differs wildly
 * (measured on two real NESDRs: dev0 [8,7,...] vs dev1 [8,7,...] off one FM
 * station despite a ~2-bit gain gap). A fine per-band shape does NOT survive that
 * gain gap; the dominant-band SET does.
 */
function topBands(spectrum: number[], k: number): number[] {
  const ranked = spectrum
    .map((v, i) => [i, v] as [number, number])
    .sort((a, b) => b[1] - a[1]);
  return ranked
    .slice(0, Math.min(k, ranked.length))
    .map((x) => x[0])
    .sort((a, b) => a - b);
}

/**
 * Encode the top-K band indices into a stable integer witness: 8 bits per band
 * (supports up to 256 bands), ascending. Decodable so agreement can be a SET
 * overlap rather than brittle exact equality.
 */
function encodeBands(bands: number[]): number {
  let w = 0;
  for (const b of bands) w = (w << 8) | (b & 0xff);
  return w >>> 0;
}

function decodeBands(witness: number, k: number): number[] {
  const out: number[] = [];
  for (let i = 0; i < k; i++) out.unshift((witness >>> (8 * i)) & 0xff);
  return out;
}

/**
 * Extract a deterministic witness from a u8 IQ capture (spectral fingerprint).
 */
export function extractWitness(
  iqU8: Uint8Array,
  opts: WitnessOptions = {},
): WitnessResult {
  const { spectrum, concentration } = bandSpectrum(iqU8, opts);
  const top = topBands(spectrum, WITNESS_TOP_K);
  const witness = encodeBands(top);
  return { witness, confidence: concentration, spectrum };
}

/**
 * Extract a witness from a decoded rtl_433 field (alternate source). Both
 * stations independently decode the same on-air device frame and recover the
 * same id/value, which we fold into the same 16-bit witness space.
 */
export function extractWitnessFromRtl433(
  event: Rtl433Event,
  field: 'id' | 'channel' = 'id',
): WitnessResult {
  const value = field === 'channel' ? event.channel : event.id;
  const descriptor = `SKYMESH-RTL433/1 model=${event.model} field=${field} value=${value ?? ''}`;
  const witness = crc16CcittAscii(descriptor);
  // A decoded field that exists is a clean lock; an absent field is low-confidence.
  const confidence = value !== null ? 1 : 0;
  return { witness, confidence, spectrum: [] };
}

/**
 * Do two independently-derived witnesses agree? Spectral witnesses encode the
 * top-K dominant bands; two receivers of the same sky agree when their dominant
 * band SETS overlap in at least `WITNESS_MIN_SHARED` bands. This is resilient: it
 * survives the noisiest band flipping between receivers (the real gain gap) while
 * still rejecting a different signal or a tampered witness. Decoded-field
 * witnesses (rtl_433 / RDS) are exact integers and match at shared == K.
 */
export function witnessAgree(a: number, b: number, minShared = WITNESS_MIN_SHARED): boolean {
  if ((a >>> 0) === (b >>> 0)) return true; // exact (decoded-field or identical spectra)
  // A non-positive request means "use the resilient default", never "agree on zero
  // shared bands" (that would make every witness agree -- the opposite of intent).
  const need = minShared > 0 ? minShared : WITNESS_MIN_SHARED;
  const sa = new Set(decodeBands(a >>> 0, WITNESS_TOP_K));
  const sb = decodeBands(b >>> 0, WITNESS_TOP_K);
  let shared = 0;
  for (const x of sb) if (sa.has(x)) shared++;
  return shared >= need;
}
