// @a0n/gnosis/rtlsdr-mock-sim/skymesh-route.mts
#!/usr/bin/env node
/**
 * Skymesh routing bridge.
 *
 * The HTTP bridge an aeon-shell `MeshRhizomeTransport` calls to resolve a rhizome
 * content-address to a mesh holder and route a payload to it via the cache-key
 * teleport.
 *
 * The unification this server implements:
 *   - A rhizome content-address IS a teleport key. Its content-hash folds to the
 *     amplituhedron `prefix_hash` (see `addressToPrefixHash`). For the canonical
 *     demo the address maps to the live FOIL key broadcastSymbol=66 / FOIL
 *     operand=7 (66 XOR 7 = 69), so the existing protocol69 admission proves the
 *     route end to end.
 *   - DID/UCAN is the protocol69 sovereign gate (fail-closed on drift): an admit
 *     that does not project to 69 is `denied`, never silently downgraded.
 *   - Gossip is the resolver: a holder is found by asking configured peers and/or
 *     a foil/fat-station whether they hold the key. With nothing configured, the
 *     bridge resolves honestly against a local in-memory registry that the caller
 *     seeds via POST /skymesh/seed.
 *
 * Endpoints (CORS open so the shell can call cross-origin):
 *   POST /skymesh/resolve  {address}            -> {addr, resolved, holder?, prefixHash?}
 *   POST /skymesh/route    {address, payload?}  -> teleport HIT/miss/denied result
 *   POST /skymesh/seed     {address, holder?}   -> seed the in-memory registry
 *   GET  /healthz                               -> "ok"
 *   OPTIONS *                                   -> 204
 *
 * Honest about hardware/services: the teleport HIT on /skymesh/route requires a
 * LIVE foil/fat-station to admit the protocol69 envelope and replay a frozen
 * amplituhedron volume. With no station configured, /skymesh/resolve still works
 * against the in-memory registry, and /skymesh/route reports `delivered:false,
 * reason:"miss"` (it does not fabricate a teleport).
 *
 * Mirrors skymesh-serve.mts: node:http only, no extra deps.
 */
import { createServer, type IncomingMessage, type ServerResponse } from 'node:http';
import { resolve as resolvePath } from 'node:path';
import { performance } from 'node:perf_hooks';
import {
  PROTOCOL69_BROADCAST_SYMBOL,
  PROTOCOL69_LOCAL_FOIL_OPERAND,
  PROTOCOL69_EXPECTED_PROJECTION,
  protocol69Projection,
  type Protocol69ProjectionEnvelope,
} from '../src/protocol69.ts';

// --------------------------------------------------------------------------
// CLI helpers (mirror skymesh-serve.mts).
// --------------------------------------------------------------------------

function argValue(name: string, def: string): string {
  const i = process.argv.indexOf(name);
  if (i === -1 || i + 1 >= process.argv.length) return def;
  return process.argv[i + 1]!;
}

// --------------------------------------------------------------------------
// Address -> prefix_hash mapping.
//
// A rhizome content-address (a CID-like string, a graphCid, an aeon:// URI, or
// any opaque content-hash) is folded deterministically into the amplituhedron
// (prefix_hash, prefix_len) cache-key coordinates.
//
//   - prefix_hash: a 24-bit fold of the address bytes (FNV-1a, masked to a safe
//     non-negative int). Deterministic and stable for one address; different
//     addresses overwhelmingly differ.
//   - prefix_len: the canonical FOIL operand (7) -- the protocol69 local operand
//     that XORs the broadcast symbol to 69. We keep it fixed at 7 so every
//     address lives on the same FOIL fiber; the address selects the *volume*
//     (prefix_hash), the protocol selects the *gate* (prefix_len).
// --------------------------------------------------------------------------

/** The canonical demo address always maps onto the live FOIL key 66/7. */
export const CANONICAL_DEMO_PREFIX_HASH = PROTOCOL69_BROADCAST_SYMBOL; // 66
export const CANONICAL_PREFIX_LEN = PROTOCOL69_LOCAL_FOIL_OPERAND; // 7

const FNV_OFFSET_BASIS = 0x811c9dc5;
const FNV_PRIME = 0x01000193;
/** Mask to a 24-bit non-negative int (well within Number.MAX_SAFE_INTEGER). */
const PREFIX_HASH_MASK = 0x00ff_ffff;

/**
 * Fold a rhizome content-address to (prefixHash, prefixLen). Pure, total,
 * deterministic. The canonical demo address(es) map to the live FOIL key 66/7 so
 * the existing protocol69 admission proves a real route; every other address
 * folds via FNV-1a to a stable 24-bit prefix_hash on the same FOIL fiber.
 */
export function addressToPrefixHash(address: string): {
  readonly prefixHash: number;
  readonly prefixLen: number;
} {
  const normalized = address.trim();
  // The canonical demo aliases collapse onto the proven live key 66/7.
  if (isCanonicalDemoAddress(normalized)) {
    return { prefixHash: CANONICAL_DEMO_PREFIX_HASH, prefixLen: CANONICAL_PREFIX_LEN };
  }
  let hash = FNV_OFFSET_BASIS;
  for (let i = 0; i < normalized.length; i++) {
    hash ^= normalized.charCodeAt(i) & 0xff;
    // Mix higher code-unit byte too so multi-byte chars participate.
    hash = Math.imul(hash, FNV_PRIME);
    hash ^= (normalized.charCodeAt(i) >> 8) & 0xff;
    hash = Math.imul(hash, FNV_PRIME);
  }
  const prefixHash = (hash >>> 0) & PREFIX_HASH_MASK;
  return { prefixHash, prefixLen: CANONICAL_PREFIX_LEN };
}

/**
 * The canonical demo address: the literal string "66/7", "canonical", or any
 * address that already encodes the live FOIL key. This is the single address the
 * live foil/fat-station can admit + replay without a fresh amplituhedron capture
 * for some arbitrary prefix_hash.
 */
export function isCanonicalDemoAddress(address: string): boolean {
  const a = address.trim().toLowerCase();
  return (
    a === 'canonical' ||
    a === '66/7' ||
    a === `${CANONICAL_DEMO_PREFIX_HASH}/${CANONICAL_PREFIX_LEN}` ||
    a === 'aeon://rhizome/canonical' ||
    a === 'gnosis.protocol69.canonical'
  );
}

// --------------------------------------------------------------------------
// Resolution: in-memory registry + live station probe.
// --------------------------------------------------------------------------

export interface SkymeshConfig {
  readonly foilUrl: string;
  /** fat-station base URL exposing POST /.aeon/teleport-admit (no trailing slash). */
  readonly stationUrl: string;
  readonly peers: readonly string[];
  /** Override the network probe (tests inject a deterministic stub). */
  readonly probeHolder?: (
    address: string,
    coords: { prefixHash: number; prefixLen: number },
  ) => Promise<string | null>;
  /** Override the teleport admit (tests inject HIT/miss/denied deterministically). */
  readonly admit?: (
    envelope: Protocol69ProjectionEnvelope,
    coords: { prefixHash: number; prefixLen: number },
  ) => Promise<AdmitResult>;
}

/** A holder is a URL (peer/station) that claims to hold the key. */
type Holder = string;

/** The in-memory demo registry: address -> holder. Seeded via POST /skymesh/seed. */
export function makeRegistry(): Map<string, Holder> {
  return new Map<string, Holder>();
}

export interface ResolveResult {
  readonly addr: string;
  readonly resolved: boolean;
  readonly holder?: string;
  readonly prefixHash?: number;
}

/**
 * Resolve a rhizome address to a holder.
 *
 * Resolution order (gossip-first, registry as the honest local fallback):
 *   1. In-memory registry (seeded for the demo / offline use).
 *   2. A live station/peer probe: ask whether it holds the key. For the demo, the
 *      canonical 66/7 key resolves true if a station URL is configured (the live
 *      foil/fat-station is the holder of the proven volume).
 */
export async function resolveAddress(
  address: string,
  registry: Map<string, Holder>,
  config: SkymeshConfig,
): Promise<ResolveResult> {
  const { prefixHash } = addressToPrefixHash(address);

  // 1. In-memory registry first (deterministic, no network).
  const seeded = registry.get(address.trim());
  if (seeded !== undefined) {
    return { addr: address, resolved: true, holder: seeded, prefixHash };
  }

  // 2. Live probe (peers / station). Injectable for tests.
  const probe = config.probeHolder ?? defaultProbeHolder;
  const holder = await probe(address, { prefixHash, prefixLen: CANONICAL_PREFIX_LEN }).catch(
    () => null,
  );
  if (holder) {
    return { addr: address, resolved: true, holder, prefixHash };
  }

  return { addr: address, resolved: false, prefixHash };
}

/**
 * Default holder probe: ask the configured peers and station whether they hold
 * the key. Honest demo rule: the canonical 66/7 key resolves to the station if a
 * station URL is configured AND it is up (we treat a reachable /healthz as "holds
 * the canonical volume", since the fat-station seeds 66/7 on boot). Non-canonical
 * keys require a peer that explicitly claims the key via GET /skymesh/holds.
 */
async function defaultProbeHolderImpl(
  address: string,
  coords: { prefixHash: number; prefixLen: number },
  config: SkymeshConfig,
): Promise<string | null> {
  const canonical = isCanonicalDemoAddress(address);

  // Ask explicit peers first: GET <peer>/skymesh/holds?prefixHash=..&prefixLen=..
  for (const peer of config.peers) {
    const held = await peerHoldsKey(peer, coords).catch(() => false);
    if (held) return peer;
  }

  // The station holds the canonical volume if it is reachable.
  if (canonical && config.stationUrl.trim() !== '') {
    const up = await stationIsUp(config.stationUrl).catch(() => false);
    if (up) return config.stationUrl;
  }

  return null;
}

// `defaultProbeHolder` closes over the live config at server start. The exported
// `resolveAddress` accepts an injected probe so tests never touch the network.
let activeConfig: SkymeshConfig | null = null;
async function defaultProbeHolder(
  address: string,
  coords: { prefixHash: number; prefixLen: number },
): Promise<string | null> {
  if (!activeConfig) return null;
  return defaultProbeHolderImpl(address, coords, activeConfig);
}

async function peerHoldsKey(
  peer: string,
  coords: { prefixHash: number; prefixLen: number },
): Promise<boolean> {
  const base = peer.replace(/\/$/, '');
  const url = `${base}/skymesh/holds?prefixHash=${coords.prefixHash}&prefixLen=${coords.prefixLen}`;
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), 2000);
  try {
    const res = await fetch(url, { method: 'GET', signal: controller.signal });
    if (!res.ok) return false;
    const body = (await res.json().catch(() => null)) as { holds?: boolean } | null;
    return body?.holds === true;
  } finally {
    clearTimeout(timer);
  }
}

async function stationIsUp(stationUrl: string): Promise<boolean> {
  const base = stationUrl.replace(/\/$/, '');
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), 2000);
  try {
    const res = await fetch(`${base}/healthz`, { method: 'GET', signal: controller.signal });
    return res.ok;
  } finally {
    clearTimeout(timer);
  }
}

// --------------------------------------------------------------------------
// Route: cache-key teleport via protocol69 admission.
// --------------------------------------------------------------------------

export type AdmitOutcome = 'hit' | 'miss' | 'denied';

export interface AdmitResult {
  readonly outcome: AdmitOutcome;
  readonly status: number;
  readonly geodesicLength?: number;
  readonly replayMicros?: number;
}

/**
 * Build the protocol69 envelope for an address. The demo maps the address onto
 * the canonical broadcast symbol 66 (FOIL operand 7 -> projection 69). A forged
 * symbol (anything but 66) projects to != 69 and is rejected by the gate. We
 * expose `broadcastSymbolOverride` so the test can drive the forged/denied path.
 */
export function buildEnvelopeForAddress(
  address: string,
  broadcastSymbolOverride?: number,
): Protocol69ProjectionEnvelope {
  const broadcastSymbol = broadcastSymbolOverride ?? PROTOCOL69_BROADCAST_SYMBOL;
  const operand = PROTOCOL69_LOCAL_FOIL_OPERAND;
  const expected = protocol69Projection(broadcastSymbol, operand);
  return {
    schemaVersion: 'gnosis.protocol69.v1',
    protocol: 'protocol69',
    wordName: 'sixty9',
    mode: 'monster-number',
    magic: 'GNOSIS-MONSTER-NUMBER/1',
    broadcastSymbol: broadcastSymbol as 66,
    localFoilOperand: operand,
    operator: 'xor',
    expectedProjection: expected as 69,
    ascii: `GNOSIS-MONSTER-NUMBER/1 skymeshRoute address=${address} broadcastSymbol=${broadcastSymbol} localFoilOperand=${operand} operator=xor expectedProjection=${expected}`,
  };
}

/**
 * Admit a cache key against a live fat-station: POST /.aeon/teleport-admit with
 * the protocol69 envelope. 200 + X-Geodesic-Length:0 -> hit; 403 -> denied
 * (drift, fail-closed); 404 -> miss (admitted but no frozen volume). Any other
 * status / network error -> miss (no station, nothing teleported).
 */
async function admitAgainstStation(
  envelope: Protocol69ProjectionEnvelope,
  config: SkymeshConfig,
): Promise<AdmitResult> {
  const base = config.stationUrl.trim();
  if (base === '') {
    // No live station: we cannot honestly teleport. Report a miss, not a fake hit.
    return { outcome: 'miss', status: 0 };
  }
  const url = `${base.replace(/\/$/, '')}/.aeon/teleport-admit`;
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), 5000);
  try {
    const res = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(envelope),
      signal: controller.signal,
    });
    if (res.status === 200) {
      const geo = Number.parseInt(res.headers.get('X-Geodesic-Length') ?? '0', 10);
      const micros = Number.parseInt(res.headers.get('X-Replay-Micros') ?? '0', 10);
      return {
        outcome: 'hit',
        status: 200,
        geodesicLength: Number.isFinite(geo) ? geo : 0,
        replayMicros: Number.isFinite(micros) ? micros : undefined,
      };
    }
    if (res.status === 403) return { outcome: 'denied', status: 403 };
    return { outcome: 'miss', status: res.status }; // 404 + anything else
  } catch {
    return { outcome: 'miss', status: 0 };
  } finally {
    clearTimeout(timer);
  }
}

export interface RouteResult {
  readonly delivered: boolean;
  readonly geodesicLength?: number;
  readonly token?: string;
  readonly replayMicros?: number;
  readonly reason?: 'miss' | 'denied';
  /** Where a /chain round-trip would go on a miss (informational). */
  readonly chainHint?: string;
}

/**
 * Route a payload to the holder of a rhizome address via the cache-key teleport.
 *
 *   - HIT (admitted + replayed locally): geodesicLength:0, a token, replayMicros.
 *   - drift/denied (projection != 69, forged key): delivered:false, reason:denied
 *     (fail-closed -- we never downgrade a denied admit to a delivery).
 *   - miss (no frozen volume / no station): delivered:false, reason:miss, plus a
 *     hint at where a /chain round-trip would fetch + freeze the volume.
 */
export async function routeAddress(
  address: string,
  config: SkymeshConfig,
  opts: { readonly broadcastSymbolOverride?: number; readonly payload?: unknown } = {},
): Promise<RouteResult> {
  const coords = addressToPrefixHash(address);
  const envelope = buildEnvelopeForAddress(address, opts.broadcastSymbolOverride);

  // Sovereign gate: if the local projection already drifts off 69, fail closed
  // before we ever touch the station (a forged symbol can never deliver).
  if (envelope.expectedProjection !== PROTOCOL69_EXPECTED_PROJECTION) {
    return { delivered: false, reason: 'denied' };
  }

  const admit = config.admit ?? ((env) => admitAgainstStation(env, config));
  const t0 = performance.now();
  const result = await admit(envelope, coords);
  const elapsedMicros = (performance.now() - t0) * 1000;

  if (result.outcome === 'hit') {
    return {
      delivered: true,
      geodesicLength: result.geodesicLength ?? 0,
      token: teleportToken(address, coords),
      replayMicros: result.replayMicros ?? Number(elapsedMicros.toFixed(2)),
    };
  }
  if (result.outcome === 'denied') {
    return { delivered: false, reason: 'denied' };
  }
  // miss
  return {
    delivered: false,
    reason: 'miss',
    chainHint:
      config.stationUrl.trim() !== ''
        ? `${config.stationUrl.replace(/\/$/, '')}/.aeon/chain (fetch + amplituhedron-capture prefix_hash=${coords.prefixHash}, then retry)`
        : 'no station configured: seed a holder (POST /skymesh/seed) or set --station-url to teleport',
  };
}

/**
 * Deterministic teleport token: the proof a holder served the cached answer for
 * this address without traversing compute space (geodesicLength:0). Stable per
 * (address, coords) so the caller can dedupe replays.
 */
export function teleportToken(
  address: string,
  coords: { prefixHash: number; prefixLen: number },
): string {
  return `teleport:${coords.prefixHash}/${coords.prefixLen}:${PROTOCOL69_EXPECTED_PROJECTION}:${address.trim()}`;
}

// --------------------------------------------------------------------------
// HTTP handlers (pure-ish: take the registry + config, return status + body).
// The tests call these directly; the server wires them to node:http.
// --------------------------------------------------------------------------

const CORS_HEADERS = {
  'Access-Control-Allow-Origin': '*',
  'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
  'Access-Control-Allow-Headers': 'Content-Type',
} as const;

export interface HandlerResponse {
  readonly status: number;
  readonly body: unknown;
}

export async function handleResolve(
  body: { address?: unknown },
  registry: Map<string, Holder>,
  config: SkymeshConfig,
): Promise<HandlerResponse> {
  const address = typeof body.address === 'string' ? body.address : '';
  if (address.trim() === '') {
    return { status: 400, body: { error: 'address required' } };
  }
  const result = await resolveAddress(address, registry, config);
  return { status: 200, body: result };
}

export async function handleRoute(
  body: { address?: unknown; payload?: unknown; broadcastSymbol?: unknown },
  config: SkymeshConfig,
): Promise<HandlerResponse> {
  const address = typeof body.address === 'string' ? body.address : '';
  if (address.trim() === '') {
    return { status: 400, body: { error: 'address required' } };
  }
  const broadcastSymbolOverride =
    typeof body.broadcastSymbol === 'number' ? body.broadcastSymbol : undefined;
  const result = await routeAddress(address, config, {
    broadcastSymbolOverride,
    payload: body.payload,
  });
  return { status: 200, body: result };
}

export function handleSeed(
  body: { address?: unknown; holder?: unknown },
  registry: Map<string, Holder>,
): HandlerResponse {
  const address = typeof body.address === 'string' ? body.address : '';
  if (address.trim() === '') {
    return { status: 400, body: { error: 'address required' } };
  }
  const holder =
    typeof body.holder === 'string' && body.holder.trim() !== ''
      ? body.holder
      : 'in-memory://skymesh-registry';
  registry.set(address.trim(), holder);
  const { prefixHash } = addressToPrefixHash(address);
  return { status: 200, body: { seeded: true, addr: address, holder, prefixHash } };
}

// --------------------------------------------------------------------------
// Server.
// --------------------------------------------------------------------------

function readJsonBody(req: IncomingMessage): Promise<Record<string, unknown>> {
  return new Promise((resolveBody) => {
    const chunks: Buffer[] = [];
    req.on('data', (c: Buffer) => chunks.push(c));
    req.on('end', () => {
      const raw = Buffer.concat(chunks).toString('utf8');
      if (raw.trim() === '') return resolveBody({});
      try {
        resolveBody(JSON.parse(raw) as Record<string, unknown>);
      } catch {
        resolveBody({});
      }
    });
    req.on('error', () => resolveBody({}));
  });
}

export interface ServerOptions {
  readonly port: number;
  readonly host: string;
  readonly config: SkymeshConfig;
}

export function startServer(opts: ServerOptions): void {
  const registry = makeRegistry();
  activeConfig = opts.config;

  const server = createServer(async (req: IncomingMessage, res: ServerResponse) => {
    const url = (req.url ?? '/').split('?')[0];

    if (req.method === 'OPTIONS') {
      res.writeHead(204, CORS_HEADERS);
      res.end();
      return;
    }

    if (url === '/healthz' && req.method === 'GET') {
      res.writeHead(200, { ...CORS_HEADERS, 'Content-Type': 'text/plain' });
      res.end('ok');
      return;
    }

    const sendJson = (r: HandlerResponse): void => {
      res.writeHead(r.status, { ...CORS_HEADERS, 'Content-Type': 'application/json' });
      res.end(JSON.stringify(r.body));
    };

    if (url === '/skymesh/resolve' && req.method === 'POST') {
      const body = await readJsonBody(req);
      sendJson(await handleResolve(body, registry, opts.config));
      return;
    }

    if (url === '/skymesh/route' && req.method === 'POST') {
      const body = await readJsonBody(req);
      sendJson(await handleRoute(body, opts.config));
      return;
    }

    if (url === '/skymesh/seed' && req.method === 'POST') {
      const body = await readJsonBody(req);
      sendJson(handleSeed(body, registry));
      return;
    }

    res.writeHead(404, { ...CORS_HEADERS, 'Content-Type': 'application/json' });
    res.end(JSON.stringify({ error: 'not_found', path: url }));
  });

  server.listen(opts.port, opts.host, () => {
    const base = `http://${opts.host}:${opts.port}`;
    console.log(`[skymesh-route] listening on ${base}`);
    console.log(`[skymesh-route]   resolve: POST ${base}/skymesh/resolve  {address}`);
    console.log(`[skymesh-route]   route:   POST ${base}/skymesh/route    {address, payload?}`);
    console.log(`[skymesh-route]   seed:    POST ${base}/skymesh/seed     {address, holder?}`);
    console.log(`[skymesh-route]   health:  GET  ${base}/healthz`);
    console.log(`[skymesh-route]   foil-url=${opts.config.foilUrl}`);
    console.log(
      `[skymesh-route]   station-url=${opts.config.stationUrl || '(none -- /skymesh/route reports miss)'}`,
    );
    console.log(
      `[skymesh-route]   peers=${opts.config.peers.length > 0 ? opts.config.peers.join(',') : '(none)'}`,
    );
  });
}

// --------------------------------------------------------------------------
// Entry point (skipped when imported, e.g. by the test).
// --------------------------------------------------------------------------

function main(): void {
  if (process.argv.includes('--help') || process.argv.includes('-h')) {
    console.log(
      [
        'skymesh-route -- resolve a rhizome address to a mesh holder + teleport-route it',
        '',
        'Usage: tsx skymesh-route.mts [--port 8731] [--station-url <url>] [--peers a,b]',
        '',
        '  --port <n>         listen port (default 8731)',
        '  --host <addr>      bind address (default 127.0.0.1)',
        '  --foil-url <url>   protocol69 foil validator (default http://127.0.0.1:8787/.aeon/protocol69)',
        '  --station-url <u>  fat-station base for POST /.aeon/teleport-admit (enables real routing)',
        '  --peers <a,b>      comma-separated peer base URLs for gossip resolution',
        '',
        'Endpoints: POST /skymesh/resolve, POST /skymesh/route, POST /skymesh/seed, GET /healthz',
        '',
        'Honest: a teleport HIT on /skymesh/route needs a LIVE fat-station (--station-url).',
        'Without one, /skymesh/resolve still works against the in-memory registry (POST',
        '/skymesh/seed) and /skymesh/route reports delivered:false, reason:"miss".',
      ].join('\n'),
    );
    return;
  }

  const port = Number.parseInt(argValue('--port', '8731'), 10);
  const host = argValue('--host', '127.0.0.1');
  const foilUrl = argValue('--foil-url', 'http://127.0.0.1:8787/.aeon/protocol69');
  const stationUrl = argValue('--station-url', '');
  const peersRaw = argValue('--peers', '');
  const peers = peersRaw
    .split(',')
    .map((p) => p.trim())
    .filter((p) => p !== '');

  startServer({ port, host, config: { foilUrl, stationUrl, peers } });
}

const isMain = (() => {
  try {
    const invoked = process.argv[1] ? resolvePath(process.argv[1]) : '';
    return invoked.endsWith('skymesh-route.mts') || invoked.endsWith('skymesh-route.js');
  } catch {
    return false;
  }
})();

if (isMain) {
  try {
    main();
  } catch (e) {
    console.error(e instanceof Error ? e.message : String(e));
    process.exitCode = 1;
  }
}
