// @a0n/gnosis/rtlsdr-mock-sim/skymesh.ts
/**
 * Skymesh protocol + shared frame schema.
 *
 * The demo: two air-gapped stations, NO network between them, both tune to the
 * SAME always-on ambient RF broadcast, both independently derive the SAME
 * witness (see skymesh-witness.ts), both admit protocol69 (66 XOR 7 = 69) and
 * serve the SAME geodesicLength:0 cached answer in lockstep.
 *
 * Honest framing: the answer ("Paris", the project's canonical qspec result) was
 * precomputed ONCE and cached. The shared witness from the sky teleports WHICH
 * cached answer to serve -- it does not recompute it. A wrong/absent/tampered
 * witness yields NO answer (fail-closed).
 */
import {
  PROTOCOL69_BROADCAST_SYMBOL,
  PROTOCOL69_LOCAL_FOIL_OPERAND,
  PROTOCOL69_EXPECTED_PROJECTION,
  protocol69Projection,
} from '../src/protocol69';
import { witnessAgree } from './skymesh-witness';
import { mkdirSync, writeFileSync } from 'node:fs';
import { dirname, resolve } from 'node:path';

export const SKYMESH_SCHEMA_VERSION = 'gnosis.skymesh.v1';
export const SKYMESH_CERTIFICATE_VERSION = 'gnosis.skymesh-certificate.v1';

/** The precomputed cached compute result the witness teleports. */
export const SKYMESH_CACHED_ANSWER = 'Paris';

/** The geodesicLength of a cache hit: 0 (no traversal of compute space). */
export const SKYMESH_GEODESIC_LENGTH = 0;

export const DEFAULT_SKYMESH_FRAME_PATH = '.qa-artifacts/skymesh/frame.json';

export type StationState =
  | 'LISTENING'
  | 'WITNESS_LOCKED'
  | 'TELEPORTED'
  | 'DENIED';

export interface SkymeshStation {
  readonly id: string;
  readonly witness: number;
  readonly witnessLocked: boolean;
  readonly projection: number;
  readonly geodesicLength: number;
  readonly token: string | null;
  readonly state: StationState;
  readonly spectrum: number[];
}

export interface SkymeshFrame {
  readonly schemaVersion: typeof SKYMESH_SCHEMA_VERSION;
  readonly stations: SkymeshStation[];
  readonly airGap: {
    readonly linkBetweenStations: false;
    readonly proof: string;
  };
  readonly match: boolean;
}

/** Standard air-gap attestation string for the frame/certificate. */
export const AIR_GAP_PROOF =
  'no network link between stations; each station derives its witness LOCALLY ' +
  'from its own antenna capture; the shared ambient sky is the only common ' +
  'channel; no station data crosses to the other.';

/**
 * Map a witness to a protocol69 admission decision for one station.
 *
 * Admission gate: the station must (a) have locked a witness off the air, and
 * (b) that witness must AGREE with the reference witness shared by the sky
 * (i.e. both stations saw the same ambient signal). On admission the station
 * projects the CANONICAL envelope (broadcastSymbol 66) XOR the local foil
 * operand 7 -> 69 and serves the cached answer. Otherwise: no token, DENIED.
 *
 * `foilValidated` lets the orchestrator fold in a live foil-node verdict: if a
 * real protocol69 node is reachable and rejected the envelope, we fail closed.
 */
export interface AdmissionInput {
  /** This station's independently-derived witness. */
  readonly witness: number;
  /** Did this station lock a witness at all (confidence over threshold)? */
  readonly witnessLocked: boolean;
  /** The reference witness from the shared sky (the other station's lock). */
  readonly referenceWitness: number;
  /** Witness equality tolerance (0 for spectral fingerprints). */
  readonly tolerance?: number;
  /** Broadcast symbol to project (canonical 66 by default). */
  readonly broadcastSymbol?: number;
  /**
   * Optional live foil verdict. undefined = no live node consulted (rehearsal);
   * true = live node validated; false = live node rejected -> fail closed.
   */
  readonly foilValidated?: boolean;
}

export interface AdmissionResult {
  readonly admitted: boolean;
  readonly projection: number;
  readonly geodesicLength: number;
  readonly token: string | null;
  readonly state: StationState;
}

/**
 * Handles the gnosis admit Station workflow.
 */
export function admitStation(input: AdmissionInput): AdmissionResult {
  const broadcastSymbol = input.broadcastSymbol ?? PROTOCOL69_BROADCAST_SYMBOL;
  const projection = protocol69Projection(
    broadcastSymbol,
    PROTOCOL69_LOCAL_FOIL_OPERAND,
  );
  const projectionOk = projection === PROTOCOL69_EXPECTED_PROJECTION;

  // No lock -> still LISTENING (no witness off the air yet).
  if (!input.witnessLocked) {
    return {
      admitted: false,
      projection,
      geodesicLength: SKYMESH_GEODESIC_LENGTH,
      token: null,
      state: 'LISTENING',
    };
  }

  const sharedSky = witnessAgree(
    input.witness,
    input.referenceWitness,
    input.tolerance ?? 0,
  );

  // Fail closed: a live foil verdict of false denies regardless of the witness.
  const foilDenied = input.foilValidated === false;

  const admitted = sharedSky && projectionOk && !foilDenied;
  if (!admitted) {
    // Locked a witness but it does not match the sky (or projection/foil failed):
    // DENIED, no token. Tamper and novel-witness both land here.
    return {
      admitted: false,
      projection,
      geodesicLength: SKYMESH_GEODESIC_LENGTH,
      token: null,
      state: 'DENIED',
    };
  }

  return {
    admitted: true,
    projection,
    geodesicLength: SKYMESH_GEODESIC_LENGTH,
    token: SKYMESH_CACHED_ANSWER,
    state: 'TELEPORTED',
  };
}

/**
 * Build a station record from its own witness lock + the admission verdict.
 * The state distinguishes WITNESS_LOCKED (locked but not yet/never admitted)
 * from TELEPORTED/DENIED/LISTENING.
 */
export function buildStation(args: {
  readonly id: string;
  readonly witness: number;
  readonly witnessLocked: boolean;
  readonly spectrum: number[];
  readonly admission: AdmissionResult;
}): SkymeshStation {
  // If the station locked a witness but the admission only reached the lock
  // (e.g. no reference yet), surface WITNESS_LOCKED rather than the raw verdict.
  const state: StationState =
    args.witnessLocked && args.admission.state === 'LISTENING'
      ? 'WITNESS_LOCKED'
      : args.admission.state;
  return {
    id: args.id,
    witness: args.witness,
    witnessLocked: args.witnessLocked,
    projection: args.admission.projection,
    geodesicLength: args.admission.geodesicLength,
    token: args.admission.token,
    state,
    spectrum: args.spectrum,
  };
}

/**
 * Assemble the gnosis.skymesh.v1 frame from station records. `match` is true
 * iff every station teleported the SAME non-null token in lockstep.
 */
export function buildSkymeshFrame(stations: SkymeshStation[]): SkymeshFrame {
  const tokens = stations.map((s) => s.token);
  const allTeleported =
    stations.length > 0 &&
    stations.every((s) => s.state === 'TELEPORTED' && s.token !== null);
  const allEqual =
    allTeleported && tokens.every((t) => t === tokens[0]);
  return {
    schemaVersion: SKYMESH_SCHEMA_VERSION,
    stations,
    airGap: {
      linkBetweenStations: false,
      proof: AIR_GAP_PROOF,
    },
    match: allEqual,
  };
}

/** Write the skymesh frame as JSON (default .qa-artifacts/skymesh/frame.json). */
export function writeSkymeshFrame(
  path: string,
  frame: SkymeshFrame,
): string {
  const target = resolve(path);
  mkdirSync(dirname(target), { recursive: true });
  writeFileSync(target, JSON.stringify(frame, null, 2));
  return target;
}
