// @a0n/gnosis/rtlsdr-mock-sim/synthetic.ts
import { appendComplexU8 } from './iq';

export interface SyntheticFmNwrLikeOptions {
  readonly sampleRateHz: number;
  readonly durationSec: number;
  /** RMS of complex AWGN relative to unit-scale constellation (0 = off). */
  readonly noiseRms: number;
  /** IF error applied at complex baseband (Hz). Simulates PPM / tuning miss. */
  readonly ifErrorHz: number;
  /** FM deviation scale for embedded "audio" tone (Hz peak). */
  readonly fmDeviationHz: number;
  /** Tone mixed as message before FM (Hz). */
  readonly messageToneHz: number;
}

function wrapAngle(rad: number): number {
  const twoPi = Math.PI * 2;
  let x = rad % twoPi;
  if (x > Math.PI) x -= twoPi;
  if (x < -Math.PI) x += twoPi;
  return x;
}

/** Box–Muller pair, standard normal. */
function gaussianPair(): [number, number] {
  const u = Math.max(Number.EPSILON, Math.random());
  const v = Math.max(Number.EPSILON, Math.random());
  return [Math.sqrt(-2 * Math.log(u)) * Math.cos(Math.PI * 2 * v), Math.sqrt(-2 * Math.log(u)) * Math.sin(Math.PI * 2 * v)];
}

/**
 * Generates synthetic interleaved u8 IQ: FM-like phase modulation of a carrier
 * with a sinusoidal message (NWR voice band metaphor) plus optional AWGN.
 */
export function generateSyntheticFmNwrLikeIqU8(
  opts: SyntheticFmNwrLikeOptions,
): Uint8Array {
  const n = Math.max(1, Math.floor(opts.sampleRateHz * opts.durationSec));
  const buf: number[] = [];
  let phase = 0;
  const kIf = (Math.PI * 2 * opts.ifErrorHz) / opts.sampleRateHz;
  const kMsg = (Math.PI * 2 * opts.messageToneHz) / opts.sampleRateHz;
  const kDev = (Math.PI * 2 * opts.fmDeviationHz) / opts.sampleRateHz;
  for (let i = 0; i < n; i++) {
    const m = Math.sin(kMsg * i);
    phase = wrapAngle(phase + kIf + kDev * m);
    const re = Math.cos(phase);
    const im = Math.sin(phase);
    const nr = opts.noiseRms;
    if (nr <= 0) {
      appendComplexU8(buf, re, im);
    } else {
      const [g0, g1] = gaussianPair();
      appendComplexU8(buf, re + nr * g0, im + nr * g1);
    }
  }
  return Uint8Array.from(buf);
}

export interface SyntheticCwSpinOptions {
  readonly sampleRateHz: number;
  readonly durationSec: number;
  readonly noiseRms: number;
  /** Constant baseband spin (Hz), e.g. PPM mistune — no FM sidebands. */
  readonly spinHz: number;
}

/**
 * Pure complex exponential at `spinHz` (+ optional AWGN). Use for **verifiable**
 * IF estimation in TS; real NWR/FM belongs in `generateSyntheticFmNwrLikeIqU8`.
 */
export function generateSyntheticCwSpinIqU8(opts: SyntheticCwSpinOptions): Uint8Array {
  const n = Math.max(1, Math.floor(opts.sampleRateHz * opts.durationSec));
  const buf: number[] = [];
  const w = (Math.PI * 2 * opts.spinHz) / opts.sampleRateHz;
  for (let i = 0; i < n; i++) {
    const re = Math.cos(w * i);
    const im = Math.sin(w * i);
    const nr = opts.noiseRms;
    if (nr <= 0) {
      appendComplexU8(buf, re, im);
    } else {
      const [g0, g1] = gaussianPair();
      appendComplexU8(buf, re + nr * g0, im + nr * g1);
    }
  }
  return Uint8Array.from(buf);
}
