// @a0n/.ignored_gnosis/forkjoin-app/client/App.tsx
import { useEffect, useState, useRef, Suspense, lazy } from '@a0n/raect';
import { usePresence } from './hooks/usePresence';
import { useCursorTracking } from './hooks/useCursorTracking';
import { useRichPresenceTelemetry } from './hooks/useRichPresenceTelemetry';
import { CursorOverlay } from './components/Presence/CursorOverlay';
import { VisitorCount } from './components/Presence/VisitorCount';
import { LiveWord } from './components/LiveWord/LiveWord';
import AeonLogo from './components/AeonLogo';
import {
  FOUNDATION_PRODUCTS_TOP_DOWN,
  type FoundationProductId,
} from './components/stack-model';
import ecosystemManifest from './generated/ecosystem-manifest.json';
import './styles/global.css';

// Lazy-load Three.js hero canvas (biggest bundle impact)
const LazyHeroCanvas = lazy(() => import('./components/Hero3D/LazyHeroCanvas'));

// Lazy-load all below-fold sections — none are visible on initial 100vh hero
const StackSection = lazy(() =>
  import('./components/Sections/StackSection').then((m) => ({
    default: m.StackSection,
  }))
);
const TechSection = lazy(() =>
  import('./components/Sections/TechSection').then((m) => ({
    default: m.TechSection,
  }))
);
const EdgeworkSection = lazy(() =>
  import('./components/Sections/EdgeworkSection').then((m) => ({
    default: m.EdgeworkSection,
  }))
);
const NeuralSection = lazy(() =>
  import('./components/Sections/NeuralSection').then((m) => ({
    default: m.NeuralSection,
  }))
);
const FluxSection = lazy(() =>
  import('./components/Sections/FluxSection').then((m) => ({
    default: m.FluxSection,
  }))
);
const DashRelaySection = lazy(() =>
  import('./components/Sections/DashRelaySection').then((m) => ({
    default: m.DashRelaySection,
  }))
);
const RelaySection = lazy(() =>
  import('./components/Sections/RelaySection').then((m) => ({
    default: m.RelaySection,
  }))
);
const AeonSection = lazy(() =>
  import('./components/Sections/AeonSection').then((m) => ({
    default: m.AeonSection,
  }))
);
const DashSection = lazy(() =>
  import('./components/Sections/DashSection').then((m) => ({
    default: m.DashSection,
  }))
);
const AuthSection = lazy(() =>
  import('./components/Sections/AuthSection').then((m) => ({
    default: m.AuthSection,
  }))
);
const DemoSection = lazy(() =>
  import('./components/Sections/DemoSection').then((m) => ({
    default: m.DemoSection,
  }))
);
const PhilosophySection = lazy(() =>
  import('./components/Sections/PhilosophySection').then((m) => ({
    default: m.PhilosophySection,
  }))
);
const GetStartedSection = lazy(() =>
  import('./components/Sections/GetStartedSection').then((m) => ({
    default: m.GetStartedSection,
  }))
);
const EcosystemProductSection = lazy(() =>
  import('./components/Sections/EcosystemProductSection').then((m) => ({
    default: m.EcosystemProductSection,
  }))
);
const LiveSitesSection = lazy(() =>
  import('./components/Sections/LiveSitesSection').then((m) => ({
    default: m.LiveSitesSection,
  }))
);
const OpenSourceSection = lazy(() =>
  import('./components/Sections/OpenSourceSection').then((m) => ({
    default: m.OpenSourceSection,
  }))
);
const MCPEcosystemSection = lazy(() =>
  import('./components/Sections/MCPEcosystemSection').then((m) => ({
    default: m.MCPEcosystemSection,
  }))
);
const ResearchSection = lazy(() =>
  import('./components/Sections/ResearchSection').then((m) => ({
    default: m.ResearchSection,
  }))
);
const PlatformsSection = lazy(() =>
  import('./components/Sections/PlatformsSection').then((m) => ({
    default: m.PlatformsSection,
  }))
);
const LibrarySection = lazy(() =>
  import('./components/Sections/LibrarySection').then((m) => ({
    default: m.LibrarySection,
  }))
);

// Lazy-load interactive overlays (not needed for initial render)
const TextChannel = lazy(() =>
  import('./components/Channels/TextChannel').then((m) => ({
    default: m.TextChannel,
  }))
);
const VoiceChannel = lazy(() =>
  import('./components/Channels/VoiceChannel').then((m) => ({
    default: m.VoiceChannel,
  }))
);
const PresenceFlightDeck = lazy(() =>
  import('./components/Presence/PresenceFlightDeck').then((m) => ({
    default: m.PresenceFlightDeck,
  }))
);

function renderProductSection(productId: FoundationProductId) {
  switch (productId) {
    case 'edgework':
      return <EdgeworkSection key={productId} />;
    case 'neural':
      return <NeuralSection key={productId} />;
    case 'flux':
      return <FluxSection key={productId} />;
    case 'dashrelay':
      return <DashRelaySection key={productId} />;
    case 'relay':
      return <RelaySection key={productId} />;
    case 'aeon':
      return <AeonSection key={productId} />;
    case 'dash':
      return <DashSection key={productId} />;
    case 'aegis':
      return <AuthSection key={productId} />;
    default:
      return <EcosystemProductSection key={productId} productId={productId} />;
  }
}

interface AppProps {
  onReady?: () => void;
}

/**
 * Renders the App view.
 */
export default function App({ onReady }: AppProps) {
  const [isReady, setIsReady] = useState(false);
  const [heroInView, setHeroInView] = useState(true);
  const [canvasReady, setCanvasReady] = useState(false);
  const [sectionsReady, setSectionsReady] = useState(false);
  const [overlaysReady, setOverlaysReady] = useState(false);
  const heroRef = useRef<HTMLElement>(null);
  const sectionsTriggerRef = useRef<HTMLDivElement>(null);

  // Keep initial render static; only bring in the WebGL hero on intent or as a late fallback.
  useEffect(() => {
    let timeoutId: ReturnType<typeof setTimeout> | null = null;
    let didEnable = false;
    const enableCanvas = () => {
      if (didEnable) return;
      didEnable = true;
      setCanvasReady(true);
    };

    timeoutId = setTimeout(enableCanvas, 12000);
    window.addEventListener('pointermove', enableCanvas, {
      passive: true,
      once: true,
    });
    window.addEventListener('scroll', enableCanvas, {
      passive: true,
      once: true,
    });
    window.addEventListener('keydown', enableCanvas, { once: true });

    return () => {
      if (timeoutId) {
        clearTimeout(timeoutId);
      }
      window.removeEventListener('pointermove', enableCanvas);
      window.removeEventListener('scroll', enableCanvas);
      window.removeEventListener('keydown', enableCanvas);
    };
  }, []);

  // Track hero visibility - unmount hero canvas when not in view to free WebGL context
  useEffect(() => {
    if (!heroRef.current) return;

    const observer = new IntersectionObserver(
      ([entry]) => {
        setHeroInView(entry.isIntersecting);
      },
      { threshold: 0.1 }
    );

    observer.observe(heroRef.current);
    return () => observer.disconnect();
  }, []);

  // Defer below-fold section code until users approach the fold, or after fallback timeout.
  useEffect(() => {
    const trigger = sectionsTriggerRef.current;
    if (!trigger) return;

    let timeoutId: ReturnType<typeof setTimeout> | null = null;
    let hasActivated = false;
    const activateSections = () => {
      if (hasActivated) return;
      hasActivated = true;
      setSectionsReady(true);
      observer.disconnect();
      if (timeoutId) {
        clearTimeout(timeoutId);
      }
    };

    const observer = new IntersectionObserver(
      ([entry]) => {
        if (entry.isIntersecting) {
          activateSections();
        }
      },
      { threshold: 0, rootMargin: '0px 0px 320px 0px' }
    );

    observer.observe(trigger);
    timeoutId = setTimeout(activateSections, 8000);

    return () => {
      observer.disconnect();
      if (timeoutId) {
        clearTimeout(timeoutId);
      }
    };
  }, []);

  // Postpone interactive overlays and telemetry listeners until intent/idle.
  useEffect(() => {
    let timeoutId: ReturnType<typeof setTimeout> | null = null;
    let idleHandle: number | null = null;
    let activated = false;

    const activateOverlays = () => {
      if (activated) return;
      activated = true;
      setOverlaysReady(true);
    };

    const browserWindow = window as Window & {
      requestIdleCallback?: (
        callback: (deadline: IdleDeadline) => void,
        options?: IdleRequestOptions
      ) => number;
      cancelIdleCallback?: (handle: number) => void;
    };

    window.addEventListener('pointerdown', activateOverlays, {
      passive: true,
      once: true,
    });
    window.addEventListener('keydown', activateOverlays, { once: true });
    window.addEventListener('scroll', activateOverlays, {
      passive: true,
      once: true,
    });

    if (typeof browserWindow.requestIdleCallback === 'function') {
      idleHandle = browserWindow.requestIdleCallback(
        () => {
          activateOverlays();
        },
        { timeout: 12000 }
      );
    } else {
      timeoutId = setTimeout(activateOverlays, 12000);
    }

    return () => {
      window.removeEventListener('pointerdown', activateOverlays);
      window.removeEventListener('keydown', activateOverlays);
      window.removeEventListener('scroll', activateOverlays);
      if (idleHandle !== null) {
        browserWindow.cancelIdleCallback?.(idleHandle);
      }
      if (timeoutId) {
        clearTimeout(timeoutId);
      }
    };
  }, []);

  const {
    visitors,
    isConnected,
    collective,
    chatMessages,
    voiceParticipants,
    visitorId,
    visitorColor,
    sendCursor,
    sendChat,
    joinVoice,
    leaveVoice,
    sendFocusNode,
    sendSelection,
    sendTyping,
    sendScroll,
    sendViewport,
    sendInputState,
    sendEmotionState,
  } = usePresence();
  const { position: localCursor } = useCursorTracking({
    sendCursor,
    enabled: canvasReady || overlaysReady,
  });
  useRichPresenceTelemetry({
    enabled: overlaysReady,
    sendFocusNode,
    sendSelection,
    sendTyping,
    sendScroll,
    sendViewport,
    sendInputState,
    sendEmotionState,
  });

  // Hide static hero as soon as React paints — don't wait for WS
  useEffect(() => {
    if (!isReady) {
      setIsReady(true);
      onReady?.();
    }
  }, [isReady, onReady]);

  return (
    <div className="app">
      {/* Cursor overlay - shows other visitors' cursors (not our own) */}
      {overlaysReady ? (
        <CursorOverlay visitors={visitors.filter((v) => v.id !== visitorId)} />
      ) : null}

      {/* Hero Section - Full viewport Three.js visualization */}
      <section ref={heroRef} className="hero-section" aria-label="Hero">
        {/* Defer canvas until after first paint; unmount when scrolled away */}
        {canvasReady && heroInView ? (
          <Suspense
            fallback={
              <div
                style={{
                  position: 'absolute',
                  inset: 0,
                  background:
                    'linear-gradient(135deg, #0a0a0f 0%, #12121a 100%)',
                }}
              />
            }
          >
            <LazyHeroCanvas
              visitors={visitors}
              collective={collective}
              localCursor={localCursor}
            />
          </Suspense>
        ) : (
          <div
            style={{
              position: 'absolute',
              inset: 0,
              background: 'linear-gradient(135deg, #0a0a0f 0%, #12121a 100%)',
            }}
          />
        )}

        <div className="hero-shell">
          {/* Hero content overlay */}
          <div className="hero-content">
            <VisitorCount count={visitors.length} isConnected={isConnected} />

            <div className="hero-logo">
              <AeonLogo activitySignal={Math.min(1, visitors.length / 10)} />
            </div>

            <h1 className="hero-title">
              <span className="hero-title-line">Distributed.</span>
              <span className="hero-title-line">Collaborative.</span>
              <span className="hero-title-line accent">ForkJoin.</span>
            </h1>

            <p className="hero-subtitle">
              <em>Proof of Life</em> -- bottling infinity in distributed
              systems.
              <br />
              <strong>504 theorems</strong>. <strong>Zero sorry</strong>.{' '}
              <strong>{ecosystemManifest.counts.mcpServers} MCP servers</strong>.{' '}
              <strong>{ecosystemManifest.counts.openSource} open-source packages</strong>.
              <br />
              Fork, race, fold. From paper to production.
            </p>

            <div className="hero-cta">
              <a
                href="https://github.com/forkjoin-ai/"
                className="button button-secondary"
                target="_blank"
                rel="noopener noreferrer"
              >
                View on GitHub
              </a>
              <a
                href="https://docs.aeonflux.dev/"
                className="button button-secondary"
              >
                Open Docs
              </a>
              <a href="#get-started" className="button button-primary">
                Get Started
              </a>
              <a
                href="https://huggingface.co/forkjoin-ai"
                className="button button-secondary"
                target="_blank"
                rel="noopener noreferrer"
              >
                View on Hugging Face
              </a>
            </div>

            <div className="hero-explainer">
              <p className="explainer-text">
                The hero text is freely editable. Your input feeds a{' '}
                <strong>local-only</strong>, <strong>peer-to-peer</strong> model
                with on-device training (LOOM). Private but distributed, powered
                by{' '}
                <a href="#neural" className="explainer-link">
                  Neural
                </a>{' '}
                atop{' '}
                <a href="#dash" className="explainer-link">
                  Dash
                </a>
                , distributed over{' '}
                <a href="#relay" className="explainer-link">
                  Relay
                </a>
                , synced with{' '}
                <a href="#aeon" className="explainer-link">
                  Aeon
                </a>
                , secured with{' '}
                <a href="#aegis" className="explainer-link">
                  Aegis
                </a>
                , and served with{' '}
                <a href="#flux" className="explainer-link">
                  Aeon Flux
                </a>
                .
              </p>
            </div>
          </div>

          <div className="scroll-indicator" aria-hidden="true">
            <span className="scroll-text">Scroll to explore</span>
            <div className="scroll-arrow" />
          </div>
        </div>
      </section>

      <div ref={sectionsTriggerRef} aria-hidden="true" />

      {/* Content Sections — lazy-loaded, below the 100vh hero fold */}
      {sectionsReady ? (
        <Suspense fallback={null}>
          <TechSection />
          <StackSection />
          {FOUNDATION_PRODUCTS_TOP_DOWN.map((product) =>
            renderProductSection(product.id)
          )}
          <LiveSitesSection />
          <OpenSourceSection />
          <MCPEcosystemSection />
          <ResearchSection />
          <PlatformsSection />
          <LibrarySection />
          <DemoSection />
          <PhilosophySection />
          <GetStartedSection />
        </Suspense>
      ) : null}

      {/* Footer */}
      <footer className="footer">
        <div className="container">
          <div className="footer-content">
            <div className="footer-brand">
              <div
                style={{
                  display: 'flex',
                  alignItems: 'center',
                  gap: '1rem',
                  marginBottom: 'var(--space-sm)',
                }}
              >
                <div
                  style={{
                    width: '48px',
                    height: '48px',
                    color: 'var(--color-accent)',
                  }}
                >
                  <AeonLogo />
                </div>
                <h3 style={{ margin: 0 }}>ForkJoin</h3>
              </div>
              <p>
                An open-source project of{' '}
                <a
                  href="https://edgework.ai"
                  target="_blank"
                  rel="noopener noreferrer"
                >
                  edgework.ai
                </a>
                , a subsidiary of{' '}
                <a
                  href="https://affectively.ai"
                  target="_blank"
                  rel="noopener noreferrer"
                >
                  affectively.ai
                </a>
              </p>
            </div>
            <div className="footer-links">
              <a href="https://github.com/forkjoin-ai/">GitHub</a>
              <a href="https://huggingface.co/forkjoin-ai">Hugging Face</a>
              <a href="https://docs.aeonflux.dev/">Documentation</a>
              <a href="https://discord.gg/aeon">Discord</a>
            </div>
            <div className="footer-legal">
              <a href="#/privacy">Privacy</a>
              <a href="#/terms">Terms</a>
              <a href="#/cookies">Cookies</a>
            </div>
          </div>
        </div>
        <div className="footer-bottom">
          <blockquote className="footer-quote">
            "Universal intelligence should be available universally."
            <cite className="footer-attribution">
              —{' '}
              <a
                href="https://buley.fyi"
                target="_blank"
                rel="noopener noreferrer"
              >
                Taylor William Buley
              </a>
              , Architect
            </cite>
          </blockquote>
        </div>
      </footer>

      {/* Channels — lazy-loaded interactive overlays */}
      {overlaysReady ? (
        <Suspense fallback={null}>
          <TextChannel
            messages={chatMessages}
            onSend={sendChat}
            isConnected={isConnected}
            visitorColor={visitorColor}
          />
          <VoiceChannel
            participants={voiceParticipants}
            isConnected={isConnected}
            visitorId={visitorId}
            visitorColor={visitorColor}
            onJoin={joinVoice}
            onLeave={leaveVoice}
          />
          <PresenceFlightDeck
            visitors={visitors}
            localVisitorId={visitorId}
            isConnected={isConnected}
            onScrollStateChange={sendScroll}
          />
        </Suspense>
      ) : null}

      <style>{`
        .hero-section {
          position: relative;
          height: 100vh;
          min-height: 600px;
          display: flex;
          justify-content: center;
          overflow: hidden;
        }

        .hero-shell {
          position: relative;
          z-index: 10;
          width: 100%;
          max-width: 960px;
          min-height: 100%;
          display: flex;
          align-items: center;
          justify-content: center;
          padding: clamp(2.5rem, 8vh, 6rem) var(--space-xl)
            clamp(5rem, 14vh, 7rem);
        }

        .hero-content {
          position: relative;
          text-align: center;
          width: min(100%, 800px);
        }

        .hero-logo {
          width: clamp(96px, 16vh, 160px);
          height: clamp(96px, 16vh, 160px);
          margin: 0 auto clamp(1.25rem, 4vh, 2rem);
          color: var(--color-accent);
        }

        .hero-title {
          font-size: clamp(2.5rem, 8vw, 5rem);
          font-weight: 300;
          line-height: 0.96;
          letter-spacing: -0.045em;
          margin-bottom: var(--space-lg);
        }

        .hero-title-line {
          display: block;
        }

        .hero-title .accent {
          background: linear-gradient(135deg, var(--color-accent) 0%, #4488aa 100%);
          -webkit-background-clip: text;
          -webkit-text-fill-color: transparent;
          background-clip: text;
        }

        .hero-subtitle {
          font-size: 1.25rem;
          color: var(--color-text-secondary);
          margin-bottom: var(--space-xl);
          line-height: 1.8;
        }

        .hero-subtitle strong {
          color: var(--color-text);
        }

        .hero-cta {
          display: flex;
          gap: var(--space-md);
          justify-content: center;
          flex-wrap: wrap;
          margin-bottom: var(--space-lg);
        }

        .hero-explainer {
          text-align: center;
          max-width: 600px;
          margin: 0 auto;
        }

        .explainer-text {
          font-size: 0.9rem;
          color: var(--color-text-secondary);
          line-height: 1.6;
          margin: 0;
        }

        .explainer-link {
          color: var(--color-accent);
          text-decoration: none;
          font-weight: 500;
          transition: color var(--transition-fast);
        }

        .explainer-link:hover {
          color: var(--color-accent-glow);
          text-decoration: underline;
        }

        .scroll-indicator {
          position: absolute;
          left: 50%;
          bottom: max(var(--space-md), env(safe-area-inset-bottom));
          transform: translateX(-50%);
          display: flex;
          flex-direction: column;
          align-items: center;
          gap: var(--space-xs);
          opacity: 0.5;
        }

        .scroll-text {
          font-size: 0.75rem;
          text-transform: uppercase;
          letter-spacing: 0.1em;
          color: var(--color-text-muted);
        }

        .scroll-arrow {
          width: 24px;
          height: 24px;
          border-right: 2px solid var(--color-text-muted);
          border-bottom: 2px solid var(--color-text-muted);
          transform: rotate(45deg);
          animation: scroll-bounce 2s ease-in-out infinite;
        }

        @keyframes scroll-bounce {
          0%, 100% { transform: rotate(45deg) translate(0, 0); }
          50% { transform: rotate(45deg) translate(4px, 4px); }
        }

        @media (max-height: 760px) {
          .hero-shell {
            padding-top: var(--space-xl);
            padding-bottom: clamp(4rem, 12vh, 5.5rem);
          }

          .hero-title {
            margin-bottom: var(--space-md);
          }

          .hero-subtitle {
            font-size: 1.1rem;
            line-height: 1.65;
            margin-bottom: var(--space-lg);
          }

          .hero-cta {
            margin-bottom: var(--space-md);
          }

          .explainer-text {
            font-size: 0.84rem;
            line-height: 1.5;
          }
        }

        .footer {
          background: var(--color-bg-secondary);
          border-top: 1px solid var(--color-bg-tertiary);
          padding: var(--space-3xl) 0;
        }

        .footer-content {
          display: flex;
          justify-content: space-between;
          align-items: flex-start;
          flex-wrap: wrap;
          gap: var(--space-xl);
          margin-bottom: var(--space-xl);
        }

        .footer-brand h3 {
          margin-bottom: var(--space-sm);
        }

        .footer-brand p {
          font-size: 0.875rem;
        }

        .footer-links {
          display: flex;
          gap: var(--space-lg);
        }

        .footer-links a {
          color: var(--color-text-secondary);
          font-size: 0.875rem;
        }

        .footer-links a:hover {
          color: var(--color-accent);
        }

        .footer-legal {
          display: flex;
          gap: var(--space-md);
          margin-top: var(--space-md);
        }

        .footer-legal a {
          color: var(--color-text-muted);
          font-size: 0.75rem;
        }

        .footer-legal a:hover {
          color: var(--color-text-secondary);
        }

        .footer-bottom {
          width: 100%;
          padding: var(--space-lg) var(--space-xl);
          border-top: 1px solid var(--color-bg-tertiary);
          margin-top: var(--space-xl);
        }

        .footer-quote {
          font-size: 1rem;
          font-style: italic;
          text-align: center;
          max-width: none;
          margin: 0;
          color: var(--color-text-secondary);
          border: none;
          padding: 0;
        }

        .footer-attribution {
          display: block;
          font-size: 0.75rem;
          margin-top: var(--space-sm);
          font-style: normal;
        }

        .footer-attribution a {
          color: var(--color-accent);
        }

        .footer-attribution a:hover {
          text-decoration: underline;
        }
      `}</style>
    </div>
  );
}
