// @a0n/.ignored_gnosis/forkjoin-app/client/docsLinks.ts
import {
  DOCS_HOME_SLUG,
  DOCS_LIBRARY_SLUG,
  DOCS_PRODUCTS_SLUG,
  DOCS_QUICKSTART_SLUG,
  buildDocsPageUrl,
  getDocsTrack,
  type DocsTrackId,
} from '@a0n/shared-utils/docsManifest';

export const FOUNDATION_DOCS_HOME_URL = buildDocsPageUrl(DOCS_HOME_SLUG);
export const FOUNDATION_DOCS_QUICKSTART_URL =
  buildDocsPageUrl(DOCS_QUICKSTART_SLUG);
export const FOUNDATION_DOCS_PRODUCTS_URL =
  buildDocsPageUrl(DOCS_PRODUCTS_SLUG);
export const FOUNDATION_DOCS_LIBRARY_URL = buildDocsPageUrl(DOCS_LIBRARY_SLUG);

/**
 * Handles the gnosis get Foundation Docs Track Url workflow.
 */
export function getFoundationDocsTrackUrl(trackId: DocsTrackId): string {
  const track = getDocsTrack(trackId);
  return buildDocsPageUrl(track?.docsSlug ?? DOCS_HOME_SLUG);
}

/**
 * Handles the gnosis get Foundation Api Track Url workflow.
 */
export function getFoundationApiTrackUrl(trackId: DocsTrackId): string | null {
  const track = getDocsTrack(trackId);
  return track?.apiSlug ? buildDocsPageUrl(track.apiSlug) : null;
}

/**
 * Handles the gnosis get Foundation Track Book Url workflow.
 */
export function getFoundationTrackBookUrl(trackId: DocsTrackId): string | null {
  const track = getDocsTrack(trackId);
  const ebookSlug = track?.ebookSlugs[0];
  return ebookSlug ? buildDocsPageUrl(ebookSlug) : null;
}
