// @a0n/gnosis/ts/DarkDeceptaconTransformer.ts
/**
 * DarkDeceptaconTransformer.ts
 * 
 * A faithful cross-axis, multi-dimensional execution of the Dark Deceptacon Transformer 
 * demonstrating mathematically identical zero-loss algebra mapping between a dense
 * topology and an optimized factorized topology.
 */

// A naive 2D matrix class for faithful mathematical representation
export class Matrix {
  constructor(public rows: number, public cols: number, public data: number[]) {}

  static create(rows: number, cols: number, initializer: (r: number, c: number) => number): Matrix {
    const data = new Array(rows * cols);
    for (let i = 0; i < rows; i++) {
      for (let j = 0; j < cols; j++) {
        data[i * cols + j] = initializer(i, j);
      }
    }
    return new Matrix(rows, cols, data);
  }

  get(r: number, c: number): number {
    return this.data[r * this.cols + c];
  }

  set(r: number, c: number, v: number): void {
    this.data[r * this.cols + c] = v;
  }

  mul(other: Matrix): Matrix {
    if (this.cols !== other.rows) throw new Error("Matrix dimensions mismatch for mul");
    return Matrix.create(this.rows, other.cols, (r, c) => {
      let sum = 0;
      for (let k = 0; k < this.cols; k++) {
        sum += this.get(r, k) * other.get(k, c);
      }
      return sum;
    });
  }

  transpose(): Matrix {
    return Matrix.create(this.cols, this.rows, (r, c) => this.get(c, r));
  }

  // Row permutation corresponding to Topological routing
  routeRows(route: number[]): Matrix {
    return Matrix.create(this.rows, this.cols, (r, c) => this.get(route[r], c));
  }

  toString(): string {
    let out = "[\n";
    for (let r = 0; r < this.rows; r++) {
      let row = "  [ ";
      for (let c = 0; c < this.cols; c++) {
        row += this.get(r, c).toFixed(2).padStart(6, " ") + " ";
      }
      out += row + "]\n";
    }
    return out + "]";
  }
}

// Ensure execution environment provides standard definitions
export const SEQ_LEN = 4;
export const DIM = 2;

export interface AttentionWeights {
  W_q: Matrix;
  W_k: Matrix;
  W_v: Matrix;
  W_o: Matrix;
}

export interface FFNFactorization {
  dense: Matrix;
  R: Matrix;   // Using orthogonal R directly here for simplicity
  Lambda: Matrix;
}

export interface TransformerLayer {
  route: number[];
  unroute: number[]; // the inverse of route
  attention: AttentionWeights;
  ffn: FFNFactorization;
}

// -----------------------------------------------------
// DENSE EXECUTION (unoptimized O(N^2))
// -----------------------------------------------------

/**
 * Handles the gnosis dense Attention workflow.
 */
export function denseAttention(L: TransformerLayer, X: Matrix): Matrix {
  const Q = X.mul(L.attention.W_q);
  const K = X.mul(L.attention.W_k);
  const V = X.mul(L.attention.W_v);
  
  const S = Q.mul(K.transpose());
  const O = S.mul(V);
  
  return O.mul(L.attention.W_o);
}

/**
 * Handles the gnosis dense FFN workflow.
 */
export function denseFFN(L: TransformerLayer, Y: Matrix): Matrix {
  // Y: [seq x dim], FFN.dense: [dim x dim]. Map applies monolithic dense transformation
  return Y.mul(L.ffn.dense.transpose());
}

/**
 * Handles the gnosis dense Layer Forward workflow.
 */
export function denseLayerForward(L: TransformerLayer, X: Matrix): Matrix {
  return denseFFN(L, denseAttention(L, X));
}

// -----------------------------------------------------
// OPTIMIZED EXECUTION (Factorized + Dynamically Routed)
// -----------------------------------------------------

/**
 * Handles the gnosis optimized Attention workflow.
 */
export function optimizedAttention(L: TransformerLayer, X: Matrix): Matrix {
  // 1. Shards routed into parallel processors
  const X_r = X.routeRows(L.route);
  
  // 2. Parallel rank projections 
  const Q_r = X_r.mul(L.attention.W_q);
  const K_r = X_r.mul(L.attention.W_k);
  const V_r = X_r.mul(L.attention.W_v);
  
  const S_r = Q_r.mul(K_r.transpose());
  const O_r = S_r.mul(V_r);
  
  // 3. Output projection
  const mapped = O_r.mul(L.attention.W_o);
  
  // 4. Mesh inversion aligns sequence invariant
  return mapped.routeRows(L.unroute);
}

/**
 * Handles the gnosis optimized FFN workflow.
 */
export function optimizedFFN(L: TransformerLayer, Y: Matrix): Matrix {
  // FFN mapped as R^T * Lambda * R
  const Rt = L.ffn.R.transpose();
  const F_opt = Rt.mul(L.ffn.Lambda.mul(L.ffn.R));
  return Y.mul(F_opt.transpose());
}

/**
 * Handles the gnosis optimized Layer Forward workflow.
 */
export function optimizedLayerForward(L: TransformerLayer, X: Matrix): Matrix {
  return optimizedFFN(L, optimizedAttention(L, X));
}

// -----------------------------------------------------
// TEST RUNNER
// -----------------------------------------------------

function isMain(): boolean {
  if (typeof process !== 'undefined' && process.argv && process.argv[1]) {
    // Attempt standard detection
    const execPath = process.argv[1].replace(/\\/g, '/');
    return execPath.endsWith('DarkDeceptaconTransformer.ts');
  }
  return false;
}

if (isMain()) {
  console.log("--- DarkDeceptacon Forward Pass Validation ---");
  console.log("Topology: Cross-Axis (Sequence x HiddenDim Tensor)");
  
  // Toy Input Sequence: 4 tokens x 2 hidden dim
  const X = new Matrix(SEQ_LEN, DIM, [
    1, 2,  // token 0
    3, 4,  // token 1
    5, 6,  // token 2
    7, 8   // token 3
  ]);
  
  // Toy Q,K,V,O Projections
  const W_q = new Matrix(DIM, DIM, [ 0.1, 0.2, -0.1, 0.5 ]);
  const W_k = new Matrix(DIM, DIM, [ 0.5, 0.0,  0.2, 0.8 ]);
  const W_v = new Matrix(DIM, DIM, [ 1.0, 0.0,  0.0, 1.0 ]);
  const W_o = new Matrix(DIM, DIM, [ 2.0, 1.0,  0.5, 1.0 ]);
  
  // Orthogonal Factor Matrix
  // R = [[cos 45, -sin 45], [sin 45, cos 45]]
  const cos45 = Math.cos(Math.PI / 4);
  const sin45 = Math.sin(Math.PI / 4);
  const R = new Matrix(DIM, DIM, [ cos45, -sin45, sin45, cos45 ]);
  const Lambda = new Matrix(DIM, DIM, [ 2, 0, 0, 0.5 ]);
  
  // Dense equivalent
  const Rt = R.transpose();
  const F_dense = Rt.mul(Lambda.mul(R));
  
  const layer: TransformerLayer = {
    // Permutation route maps [0,1,2,3] -> [2,0,3,1]
    route: [2, 0, 3, 1],
    // The inverse maps [0.1,2,3] -> [1,3,0,2] checking indices
    // index 0 -> 2, index 1 -> 0, index 2 -> 3, index 3 -> 1
    unroute: [1, 3, 0, 2],
    attention: { W_q, W_k, W_v, W_o },
    ffn: { dense: F_dense, R, Lambda }
  };
  
  console.log("\nInput Tensor X (Seq x Dim):");
  console.log(X.toString());
  
  console.log("\nComputing Theoretical Dense Forward...");
  const denseOut = denseLayerForward(layer, X);
  console.log(denseOut.toString());
  
  console.log("\nComputing Structured Optimized Forward (O(V log V) mesh path)...");
  const optOut = optimizedLayerForward(layer, X);
  console.log(optOut.toString());
  
  // Verification check
  let delta = 0;
  for (let i = 0; i < denseOut.data.length; i++) {
    delta += Math.abs(denseOut.data[i] - optOut.data[i]);
  }
  
  if (delta < 1e-6) {
    console.log("\n✅ SUCCESS: Zero-Loss structural parity fully verified via exact multidimensional tensor execution.");
  } else {
    console.log(`\n❌ ERROR: Outputs diverged by max delta: ${delta}`);
  }
}
