// @a0n/gnosis/src/forest/compose-winners.ts
/**
 * compose-winners.ts -- Assemble optimal polyglot topology from winner map.
 *
 * Takes a .gg source and a winner map (function → language), annotates each
 * node with its winning language, and emits the composed optimal topology.
 * This is the FOLD step of Forest's convergence loop: merge all race results
 * into a single topology with language pinned per node.
 */

import { BettyCompiler } from '../betty/compiler';

/**
 * Compose a new .gg source from the original topology with language
 * annotations derived from the winner map.
 *
 * Each node that has a winner gets a `language` property set to the
 * winning language. Nodes without winners retain their original properties.
 */
export function composeWinners(
  originalSource: string,
  winners: Map<string, string>
): string {
  if (winners.size === 0) return originalSource;

  const lines = originalSource.split('\n');
  const composed: string[] = [];

  for (const line of lines) {
    let modified = line;

    // For each winner, annotate the node declaration with the winning language
    for (const [nodeId, language] of winners) {
      // Match node declarations containing this ID
      const nodePattern = new RegExp(
        `\\(${escapeRegex(nodeId)}(\\s*:\\s*[^){\\s]+)?(\\s*\\{[^}]*\\})?\\)`,
        'g'
      );

      modified = modified.replace(nodePattern, (match, label, props) => {
        const labelPart = label ?? '';
        const existingProps = props ? props.trim().slice(1, -1).trim() : '';

        // Add or update language property
        if (existingProps.includes('language:')) {
          // Replace existing language
          const updatedProps = existingProps.replace(
            /language:\s*'[^']*'/,
            `language: '${language}'`
          );
          return `(${nodeId}${labelPart} { ${updatedProps} })`;
        }

        const newProps = existingProps
          ? `${existingProps}, language: '${language}'`
          : `language: '${language}'`;
        return `(${nodeId}${labelPart} { ${newProps} })`;
      });
    }

    composed.push(modified);
  }

  return composed.join('\n');
}

function escapeRegex(str: string): string {
  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

/**
 * Generate a summary comment block for the converged topology.
 */
export function generateConvergenceHeader(
  winners: Map<string, string>,
  generations: number,
  totalRejections: number
): string {
  const languageCounts = new Map<string, number>();
  for (const lang of winners.values()) {
    languageCounts.set(lang, (languageCounts.get(lang) ?? 0) + 1);
  }

  const summary = Array.from(languageCounts.entries())
    // Law 2 (strict ordering): primary sort by count (descending), secondary by language name for determinism
    .sort((a, b) => b[1] - a[1] || a[0].localeCompare(b[0]))
    .map(([lang, count]) => `${count} won by ${lang}`)
    .join(', ');

  return [
    '// ═══════════════════════════════════════════════════════════════════════════',
    `// FOREST CONVERGED: Generation ${generations}`,
    `// ${winners.size} functions raced. ${summary}.`,
    `// ${totalRejections} total rejections across ${generations} generation(s).`,
    '// Theorem: void_walkers_converge -- same rejection history = same distribution.',
    '// ═══════════════════════════════════════════════════════════════════════════',
    '',
  ].join('\n');
}
