// @a0n/gnosis/src/forest/forrest-skyrms-bridge.ts
/**
 * forrest-skyrms-bridge.ts -- Combine Skyrms equilibrium traces with Forrest pair-gate indices.
 *
 * **Contract:** twelve distinguished nodes (slots `0..11`) map Forest node ids to `Fin twelve`.
 * From each **hard assignment** (winner per node), emit one ordered step `{ lo, hi }` per unordered
 * slot pair whose winners **differ**, scanning `lo < hi` lexicographically. Chords sit inside the
 * **`pairsIJ`** envelope matched by `Gnosis.FoilForrestWalk` / `TwelveSlotSixtySixPairsCarrier`.
 *
 * **Skyrms side:** use `bundleSkyrmsWithForrestWalk` on the final `SkyrmsResult`, or attach
 * `SkyrmsWalkHooks.onPostIteration` and fold snapshots with `forrestWalkFromSkyrmsIterationSnapshots`
 * for a **walk-over-time** (iteration-concatenated) certificate.
 */

import type { CompilerName } from './types';
import type { SkyrmsIterationSnapshot, SkyrmsResult } from './skyrms-walk';

/** Matches Lean `Fin twelve` slot count for `pairsIJ`. */
export const TWELVE_FORREST_SLOTS = 12;

/** One strict ascending pair on twelve-slot indices (Lean `StrictAscendingPair` shape). */
export interface ForrestWalkStep {
  readonly lo: number;
  readonly hi: number;
}

/** Whether `step` uses integers with `0 ≤ lo < hi < 12`. */
export function isValidForrestWalkStep(step: ForrestWalkStep): boolean {
  const { lo, hi } = step;
  if (!Number.isInteger(lo) || !Number.isInteger(hi)) return false;
  return lo >= 0 && hi < TWELVE_FORREST_SLOTS && lo < hi;
}

/** Nat chord `(lo, hi)` aligned with Lean `StrictAscendingPair.toNatChord`. */
export function forrestChord(step: ForrestWalkStep): readonly [number, number] {
  return [step.lo, step.hi];
}

/**
 * Encode one equilibrium snapshot: discordant winner pairs among twelve pinned nodes.
 *
 * @param nodesAtSlots -- length **12**; `nodesAtSlots[k]` is the Forest node id occupying slot `k`.
 */
export function encodeHardAssignmentAsForrestWalk(
  hardAssignment: ReadonlyMap<string, CompilerName>,
  nodesAtSlots: readonly string[]
): ForrestWalkStep[] {
  if (nodesAtSlots.length !== TWELVE_FORREST_SLOTS) {
    throw new Error(
      `Forrest bridge requires exactly ${TWELVE_FORREST_SLOTS} node ids (one per slot); got ${nodesAtSlots.length}`
    );
  }
  const steps: ForrestWalkStep[] = [];
  for (let lo = 0; lo < TWELVE_FORREST_SLOTS; lo++) {
    for (let hi = lo + 1; hi < TWELVE_FORREST_SLOTS; hi++) {
      const winnerLo = hardAssignment.get(nodesAtSlots[lo]!);
      const winnerHi = hardAssignment.get(nodesAtSlots[hi]!);
      if (winnerLo === undefined || winnerHi === undefined) {
        throw new Error(
          `Forrest bridge: missing hardAssignment for node at slot ${lo} or ${hi}`
        );
      }
      if (winnerLo !== winnerHi) {
        steps.push({ lo, hi });
      }
    }
  }
  return steps;
}

/** Flatten ordered chunks into one walk list. */
export function concatForrestWalks(chunks: readonly ForrestWalkStep[][]): ForrestWalkStep[] {
  const out: ForrestWalkStep[] = [];
  for (const c of chunks) {
    out.push(...c);
  }
  return out;
}

/** Concatenate per-iteration encodings in snapshot order (Skyrms time × pair lex order). */
export function forrestWalkFromSkyrmsIterationSnapshots(
  snapshots: readonly SkyrmsIterationSnapshot[],
  nodesAtSlots: readonly string[]
): ForrestWalkStep[] {
  return concatForrestWalks(
    snapshots.map((s) => encodeHardAssignmentAsForrestWalk(s.hardAssignment, nodesAtSlots))
  );
}

/** Final equilibrium only (no iteration history). */
export function bundleSkyrmsWithForrestWalk(
  skyrms: SkyrmsResult,
  nodesAtSlots: readonly string[]
): { skyrms: SkyrmsResult; forrestWalk: ForrestWalkStep[] } {
  return {
    skyrms,
    forrestWalk: encodeHardAssignmentAsForrestWalk(skyrms.hardAssignment, nodesAtSlots),
  };
}
