// @a0n/gnosis/src/forest/iterate.ts
/**
 * iterate.ts -- Run Forest convergence across multiple iterations.
 *
 * Each iteration takes the previous converged topologies as input and runs
 * Forest again. The process tracks whether the winner mix stabilizes across
 * iterations (meta-convergence) or oscillates (the sliver nodes may flip).
 *
 * Usage:
 *   gnode run src/forest/iterate.ts
 *   gnode run src/forest/iterate.ts --iterations 5
 *   gnode run src/forest/iterate.ts --verbose
 */

import { readFileSync, writeFileSync, existsSync } from 'node:fs';
import { resolve, basename, join } from 'node:path';

import {
  runForestUntilConvergence,
  DEFAULT_FOREST_CONFIG,
} from './convergence-loop';
import { generateConvergenceHeader } from './compose-winners';
import type { ForestConfig, ConvergenceCertificate } from './types';

// ═══════════════════════════════════════════════════════════════════════════════
// Types
// ═══════════════════════════════════════════════════════════════════════════════

export interface IterationResult {
  iteration: number;
  topologyName: string;
  converged: boolean;
  generations: number;
  totalRejections: number;
  winnerMix: Record<string, number>;
  nodesShifted: number;
  totalNodes: number;
}

export interface MetaConvergenceReport {
  iterations: IterationResult[];
  metaConverged: boolean;
  metaConvergedAt: number | null;
  totalRejections: number;
}

// ═══════════════════════════════════════════════════════════════════════════════
// Helpers
// ═══════════════════════════════════════════════════════════════════════════════

function winnerMixFromMap(
  winners: Map<string, string>
): Record<string, number> {
  const mix: Record<string, number> = {};
  for (const lang of winners.values()) {
    mix[lang] = (mix[lang] ?? 0) + 1;
  }
  return mix;
}

function mixesEqual(
  a: Record<string, number>,
  b: Record<string, number>
): boolean {
  const keysA = Object.keys(a).sort();
  const keysB = Object.keys(b).sort();
  if (keysA.length !== keysB.length) return false;
  for (let i = 0; i < keysA.length; i++) {
    if (keysA[i] !== keysB[i]) return false;
    if (a[keysA[i]] !== b[keysB[i]]) return false;
  }
  return true;
}

// ═══════════════════════════════════════════════════════════════════════════════
// Iteration runner
// ═══════════════════════════════════════════════════════════════════════════════

/**
 * Handles the gnosis run Iterations workflow.
 */
export function runIterations(
  topologyName: string,
  initialSource: string,
  maxIterations: number,
  config: ForestConfig
): MetaConvergenceReport {
  const results: IterationResult[] = [];
  let currentSource = initialSource;
  let prevMix: Record<string, number> | null = null;
  let prevWinners: Map<string, string> | null = null;
  let totalRejections = 0;

  for (let iter = 0; iter < maxIterations; iter++) {
    const cert = runForestUntilConvergence(currentSource, config);
    const mix = winnerMixFromMap(cert.finalWinners);

    // Count how many nodes shifted strategy from previous iteration
    let nodesShifted = 0;
    if (prevWinners) {
      for (const [fn, lang] of cert.finalWinners) {
        if (prevWinners.get(fn) !== lang) nodesShifted++;
      }
    }

    const result: IterationResult = {
      iteration: iter,
      topologyName,
      converged: cert.converged,
      generations: cert.generations,
      totalRejections: cert.totalRejections,
      winnerMix: mix,
      nodesShifted,
      totalNodes: cert.finalWinners.size,
    };
    results.push(result);
    totalRejections += cert.totalRejections;

    if (config.verbose) {
      const mixStr = Object.entries(mix)
        .sort((a, b) => b[1] - a[1])
        .map(([lang, count]) => `${lang}:${count}`)
        .join(' ');
      console.log(
        `[Iterate:${topologyName}:${iter}] gens:${cert.generations} rej:${cert.totalRejections} shifted:${nodesShifted} mix:[${mixStr}]`
      );
    }

    // Meta-convergence: did the winner MIX stabilize?
    if (prevMix && mixesEqual(prevMix, mix) && nodesShifted <= 1) {
      return {
        iterations: results,
        metaConverged: true,
        metaConvergedAt: iter,
        totalRejections,
      };
    }

    currentSource = cert.optimalSource;
    prevMix = mix;
    prevWinners = cert.finalWinners;
  }

  return {
    iterations: results,
    metaConverged: false,
    metaConvergedAt: null,
    totalRejections,
  };
}

// ═══════════════════════════════════════════════════════════════════════════════
// CLI
// ═══════════════════════════════════════════════════════════════════════════════

if (
  typeof process !== 'undefined' &&
  process.argv[1] &&
  process.argv[1].includes('iterate')
) {
  const maxIterations = parseInt(
    process.argv.find((a) => a.startsWith('--iterations='))?.split('=')[1] ??
      '5'
  );
  const verbose = process.argv.includes('--verbose');

  const GNOSIS_ROOT = resolve(__dirname, '..', '..');
  const config: ForestConfig = {
    maxGenerations: 50,
    compilers: ['betty', 'betti', 'aeon-logic'],
    raceTimeoutMs: 10000,
    verbose,
  };

  const topologies: [string, string][] = [
    ['betti', join(GNOSIS_ROOT, 'betti.gg')],
    ['franky', join(GNOSIS_ROOT, 'examples/franky.gg')],
    ['beckett', join(GNOSIS_ROOT, 'examples/beckett.gg')],
  ];

  console.log(
    `\nForest Meta-Convergence: ${maxIterations} iterations, ${config.compilers.length} real compilers\n`
  );

  for (const [name, filePath] of topologies) {
    if (!existsSync(filePath)) {
      console.log(`Skipping ${name} (file not found)`);
      continue;
    }

    const source = readFileSync(filePath, 'utf-8');
    const report = runIterations(name, source, maxIterations, config);

    console.log(`\n${'─'.repeat(60)}`);
    console.log(
      `${name}: meta-converged=${report.metaConverged} at=${
        report.metaConvergedAt ?? 'N/A'
      } total_rejections=${report.totalRejections}`
    );

    for (const r of report.iterations) {
      const mixStr = Object.entries(r.winnerMix)
        .sort((a, b) => b[1] - a[1])
        .map(([lang, count]) => `${lang}:${count}`)
        .join(' ');
      console.log(
        `  iter ${r.iteration}: gens=${r.generations} rej=${r.totalRejections} shifted=${r.nodesShifted}/${r.totalNodes} [${mixStr}]`
      );
    }

    // Write final output
    const lastResult = report.iterations[report.iterations.length - 1];
    if (lastResult) {
      const outPath = join(GNOSIS_ROOT, 'examples', `${name}-iterated.gg`);
      // Re-run one more time to get the actual source
      const finalCert = runForestUntilConvergence(
        readFileSync(filePath, 'utf-8'),
        { ...config, verbose: false }
      );
      const header = generateConvergenceHeader(
        finalCert.finalWinners,
        finalCert.generations,
        report.totalRejections
      );
      writeFileSync(outPath, header + finalCert.optimalSource);
      console.log(`  written: ${outPath}`);
    }
  }
}
