// @a0n/gnosis/src/forest/types.ts
/**
 * types.ts -- Shared types for the Forest convergence engine.
 *
 * 14 real compilers. Every call is real. Best wins.
 */

/** In-process TypeScript compilers */
export type InProcessCompiler = 'betty' | 'betti' | 'aeon-logic';

/** Subprocess compilers (real binaries in other languages) */
export type SubprocessCompiler =
  | 'php'
  | 'rust'
  | 'java'
  | 'go'
  | 'python'
  | 'kotlin'
  | 'luajit'
  | 'ruby'
  | 'lua'
  | 'elixir'
  | 'swift'
  | 'cpp'
  | 'c';

export type CompilerName = InProcessCompiler | SubprocessCompiler;

export interface SubprocessCompilerConfig {
  name: SubprocessCompiler;
  command: string;
  args: (filepath: string) => string[];
}

export interface ForestConfig {
  maxGenerations: number;
  compilers: CompilerName[];
  raceTimeoutMs: number;
  verbose: boolean;
  /** Base path for polyglot binaries. Defaults to polyglot/target/release/ */
  polyglotBasePath?: string;
}

export interface GenerationState {
  generation: number;
  winners: Map<string, string>;
  rejections: Map<string, string[]>;
  timings: Map<string, Map<string, number>>;
}

export interface ConvergenceCertificate {
  converged: boolean;
  generations: number;
  finalWinners: Map<string, string>;
  totalRejections: number;
  generationHistory: GenerationState[];
  optimalSource: string;
}
