// @a0n/gnosis/mesh-local-mcp/src/mcp-http-sse-session.ts
/**
 * Minimal MCP client for HTTP+SSE transport (protocol compatible with MCP SDK SSEServerTransport).
 * Used by Vitest e2e; not a general-purpose SDK replacement.
 */

import type { Application } from 'express';
import type { Server } from 'node:http';

import {
  JSONRPC_VERSION,
  LATEST_PROTOCOL_VERSION,
} from '@modelcontextprotocol/sdk/types.js';

function splitSseBlocks(buffer: string): { complete: string[]; rest: string } {
  const parts = buffer.split('\n\n');
  const rest = parts.pop() ?? '';
  const complete = parts.filter((p) => p.trim().length > 0);
  return { complete, rest };
}

function parseBlock(block: string): { event: string; data: string } {
  let event = 'message';
  const dataLines: string[] = [];
  for (const line of block.split('\n')) {
    if (line.startsWith('event:')) {
      event = line.slice('event:'.length).trim();
    } else if (line.startsWith('data:')) {
      dataLines.push(line.slice('data:'.length).trimStart());
    }
  }
  return { event, data: dataLines.join('\n') };
}

export interface ListeningApp {
  readonly baseUrl: string;
  readonly close: () => Promise<void>;
}

/**
 * Handles the gnosis listen Express On Random Port workflow.
 */
export async function listenExpressOnRandomPort(
  app: Application,
  host: string,
): Promise<{ server: Server } & ListeningApp> {
  const server: Server = await new Promise((resolvePromise, reject) => {
    const s = app.listen(0, host, () => resolvePromise(s));
    s.on('error', reject);
  });
  const addr = server.address();
  if (!addr || typeof addr === 'string') {
    await new Promise<void>((r, j) => server.close((e) => (e ? j(e) : r())));
    throw new Error('expected inet listen address');
  }
  const baseUrl = `http://${host}:${addr.port}`;
  const close = async (): Promise<void> =>
    new Promise((resolveClose, rejectClose) => {
      server.close((err) => (err ? rejectClose(err) : resolveClose()));
    });
  return { server, baseUrl, close };
}

export interface JsonRpcResult {
  readonly jsonrpc: typeof JSONRPC_VERSION;
  readonly id: number | string;
  readonly result?: unknown;
  readonly error?: { code: number; message: string; data?: unknown };
}

export interface McpHttpSseSessionControls {
  readonly sessionId: string;
  request(id: number, payload: Record<string, unknown>): Promise<JsonRpcResult>;
  notify(payload: Record<string, unknown>): Promise<void>;
  disconnect(): void;
}

/**
 * Handles the gnosis open Mcp Http Sse Session workflow.
 */
export async function openMcpHttpSseSession(
  baseUrl: string,
  signal: AbortSignal,
): Promise<McpHttpSseSessionControls> {
  const res = await fetch(`${baseUrl}/mcp`, {
    signal,
    headers: { Accept: 'text/event-stream' },
  });
  if (!res.ok || !res.body) {
    throw new Error(`GET /mcp failed: ${res.status}`);
  }

  const reader = res.body.getReader();

  let buffer = '';
  let sessionId: string | null = null;
  const pending = new Map<number | string, { settle: (v: JsonRpcResult) => void }>();

  const settleIfResponse = (obj: unknown): void => {
    if (!obj || typeof obj !== 'object') return;
    const m = obj as Record<string, unknown>;
    const id = m.id;
    if (
      (typeof id === 'number' || typeof id === 'string') &&
      ('result' in m || 'error' in m)
    ) {
      const p = pending.get(id);
      if (p) {
        pending.delete(id);
        p.settle(m as unknown as JsonRpcResult);
      }
    }
  };

  const pumpPromise = (async (): Promise<void> => {
    const dec = new TextDecoder();
    try {
      while (!signal.aborted) {
        const { value, done } = await reader.read();
        if (done) break;
        buffer += dec.decode(value, { stream: true });
        const { complete, rest } = splitSseBlocks(buffer);
        buffer = rest;
        for (const block of complete) {
          const { event, data } = parseBlock(block);
          if (event === 'endpoint') {
            const u = new URL(data, baseUrl);
            sessionId = u.searchParams.get('sessionId');
          } else if (event === 'message') {
            try {
              settleIfResponse(JSON.parse(data) as unknown);
            } catch {
              /* ignore */
            }
          }
        }
      }
    } catch {
      /* aborted / cancelled */
    } finally {
      try {
        reader.releaseLock();
      } catch {
        /* already released */
      }
    }
  })();

  const waitPump = (): void => {
    void pumpPromise.catch(() => undefined);
  };
  waitPump();

  const waitForSessionId = async (timeoutMs = 10_000): Promise<string> => {
    const t0 = Date.now();
    while (sessionId === null && Date.now() - t0 < timeoutMs && !signal.aborted) {
      await new Promise((r) => setTimeout(r, 15));
    }
    if (!sessionId) throw new Error('timeout waiting for SSE endpoint event');
    return sessionId;
  };

  const sid = await waitForSessionId();

  const request = async (id: number, payload: Record<string, unknown>): Promise<JsonRpcResult> => {
    const resultPromise = new Promise<JsonRpcResult>((resolvePromise, reject) => {
      const t = setTimeout(() => reject(new Error(`MCP RPC id=${id} timeout`)), 20_000);
      pending.set(id, {
        settle: (v) => {
          clearTimeout(t);
          resolvePromise(v);
        },
      });
    });

    const postRes = await fetch(`${baseUrl}/messages?sessionId=${encodeURIComponent(sid)}`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ jsonrpc: JSONRPC_VERSION, id, ...payload }),
    });

    if (postRes.status !== 202) {
      pending.delete(id);
      throw new Error(`POST /messages expected 202, got ${postRes.status}: ${await postRes.text()}`);
    }

    const out = await resultPromise;
    if (out.error) {
      throw new Error(`MCP error id=${id}: ${out.error.message}`);
    }
    return out;
  };

  const notify = async (payload: Record<string, unknown>): Promise<void> => {
    const postRes = await fetch(`${baseUrl}/messages?sessionId=${encodeURIComponent(sid)}`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ jsonrpc: JSONRPC_VERSION, ...payload }),
    });
    if (postRes.status !== 202) {
      throw new Error(`POST notify expected 202, got ${postRes.status}`);
    }
  };

  return {
    sessionId: sid,
    request,
    notify,
    disconnect: () => {
      void reader.cancel().catch(() => undefined);
    },
  };
}

/**
 * Handles the gnosis mcp Initialize Payload workflow.
 */
export function mcpInitializePayload(): Record<string, unknown> {
  return {
    method: 'initialize',
    params: {
      protocolVersion: LATEST_PROTOCOL_VERSION,
      capabilities: {},
      clientInfo: { name: 'mesh-local-mcp-e2e', version: '0.0.0' },
    },
  };
}

/**
 * Handles the gnosis assert Initialize Ok workflow.
 */
export function assertInitializeOk(res: JsonRpcResult): void {
  const result = res.result as Record<string, unknown> | undefined;
  if (typeof result?.protocolVersion !== 'string' || !result.protocolVersion) {
    throw new Error('initialize missing protocolVersion');
  }
  const info = result.serverInfo as Record<string, unknown> | undefined;
  if (typeof info?.name !== 'string' || !info.name) {
    throw new Error('initialize missing serverInfo.name');
  }
}
