// @a0n/.ignored_gnosis/forkjoin-app/client/components/Channels/TextChannel.tsx
/**
 * TextChannel - Discord-style text chat
 *
 * Squelched by default. Wild when enabled.
 * Messages broadcast via WebSocket to all visitors.
 * TTS/STT via AI Gateway.
 */

// File-local Web Speech shapes (avoid colliding with global declarations)
interface BrowserSpeechRecognitionInterface extends EventTarget {
  continuous: boolean;
  interimResults: boolean;
  lang: string;
  start(): void;
  stop(): void;
  abort(): void;
  onresult: ((event: Event) => void) | null;
  onerror: ((event: Event) => void) | null;
  onend: ((event: Event) => void) | null;
}

interface BrowserSpeechRecognitionConstructor {
  new (): BrowserSpeechRecognitionInterface;
}

import { useState, useRef, useEffect, memo } from '@a0n/raect';
import {
  MessageCircle,
  X,
  Send,
  Unlock,
  Volume2,
  Mic,
  MicOff,
  Circle,
} from '@a0n/aeon-icons';

interface ChatMessage {
  id: string;
  visitorId: string;
  color: string;
  message: string;
  timestamp: number;
}

interface TextChannelProps {
  messages: ChatMessage[];
  onSend: (message: string) => void;
  isConnected: boolean;
  visitorColor: string;
}

export const TextChannel = memo(function TextChannel({
  messages,
  onSend,
  isConnected,
  visitorColor,
}: TextChannelProps) {
  const [isOpen, setIsOpen] = useState(false);
  const [input, setInput] = useState('');
  const [isEnabled, setIsEnabled] = useState(false);
  const [isListening, setIsListening] = useState(false);
  const [isSpeaking, setIsSpeaking] = useState(false);
  const messagesEndRef = useRef<HTMLDivElement>(null);
  const inputRef = useRef<HTMLInputElement>(null);
  const recognitionRef = useRef<BrowserSpeechRecognitionInterface | null>(null);

  // Auto-scroll to bottom on new messages
  useEffect(() => {
    messagesEndRef.current?.scrollIntoView({ behavior: 'smooth' });
  }, [messages]);

  // Initialize speech recognition
  useEffect(() => {
    const speechWindow = window as unknown as {
      SpeechRecognition?: BrowserSpeechRecognitionConstructor;
      webkitSpeechRecognition?: BrowserSpeechRecognitionConstructor;
    };
    const speechCtor =
      speechWindow.SpeechRecognition || speechWindow.webkitSpeechRecognition;

    if (speechCtor) {
      recognitionRef.current = new speechCtor();
      recognitionRef.current.continuous = false;
      recognitionRef.current.interimResults = false;

      recognitionRef.current.onresult = (event) => {
        const recognitionEvent = event as Event & {
          results?: {
            [index: number]: {
              [altIndex: number]: {
                transcript: string;
              };
            };
          };
        };
        const transcript =
          recognitionEvent.results?.[0]?.[0]?.transcript?.trim() || '';
        setInput(transcript);
        setIsListening(false);
      };

      recognitionRef.current.onerror = () => {
        setIsListening(false);
      };

      recognitionRef.current.onend = () => {
        setIsListening(false);
      };
    }
  }, []);

  const handleSend = () => {
    if (!input.trim() || !isEnabled) return;
    onSend(input.trim());
    setInput('');
  };

  const handleKeyDown = (e: React.KeyboardEvent) => {
    if (e.key === 'Enter' && !e.shiftKey) {
      e.preventDefault();
      handleSend();
    }
  };

  const enableChat = () => {
    setIsEnabled(true);
    setTimeout(() => inputRef.current?.focus(), 100);
  };

  const toggleListening = () => {
    if (!recognitionRef.current) return;

    if (isListening) {
      recognitionRef.current.stop();
      setIsListening(false);
    } else {
      recognitionRef.current.start();
      setIsListening(true);
    }
  };

  const speakMessage = (text: string) => {
    if ('speechSynthesis' in window && !isSpeaking) {
      setIsSpeaking(true);
      const utterance = new SpeechSynthesisUtterance(text);
      utterance.onend = () => setIsSpeaking(false);
      utterance.onerror = () => setIsSpeaking(false);
      speechSynthesis.speak(utterance);
    }
  };

  return (
    <>
      {/* Toggle Button */}
      <button
        className="channel-toggle text-toggle"
        onClick={() => setIsOpen(!isOpen)}
        aria-label="Toggle text channel"
        aria-expanded={isOpen}
      >
        <MessageCircle size={20} />
        {messages.length > 0 && !isOpen && (
          <span className="message-badge">{Math.min(messages.length, 99)}</span>
        )}
      </button>

      {/* Chat Panel */}
      {isOpen && (
        <div className="text-channel" role="dialog" aria-label="Text channel">
          <div className="channel-header">
            <h3># global-chat</h3>
            <span className="channel-status">
              <Circle
                size={8}
                fill={isConnected ? '#10b981' : '#ef4444'}
                color={isConnected ? '#10b981' : '#ef4444'}
              />
              {isConnected ? 'Live' : 'Offline'}
            </span>
            <button
              className="close-button"
              onClick={() => setIsOpen(false)}
              aria-label="Close chat"
            >
              <X size={16} />
            </button>
          </div>

          <div className="messages-container">
            {messages.length === 0 ? (
              <div className="empty-state">
                <MessageCircle size={32} strokeWidth={1} />
                <p>No messages yet</p>
                <p className="empty-hint">Be the first to say hi!</p>
              </div>
            ) : (
              messages.map((msg) => (
                <div key={msg.id} className="message">
                  <span
                    className="message-avatar"
                    style={{ backgroundColor: msg.color }}
                  />
                  <div className="message-content">
                    <span
                      className="message-sender"
                      style={{ color: msg.color }}
                    >
                      Visitor {msg.visitorId.slice(0, 6)}
                    </span>
                    <span className="message-text">{msg.message}</span>
                  </div>
                  <div className="message-actions">
                    <button
                      className="tts-button"
                      onClick={() => speakMessage(msg.message)}
                      aria-label="Speak message"
                      disabled={isSpeaking}
                    >
                      <Volume2 size={12} />
                    </button>
                    <span className="message-time">
                      {new Date(msg.timestamp).toLocaleTimeString([], {
                        hour: '2-digit',
                        minute: '2-digit',
                      })}
                    </span>
                  </div>
                </div>
              ))
            )}
            <div ref={messagesEndRef} />
          </div>

          <div className="input-container">
            {!isEnabled ? (
              <button className="enable-button" onClick={enableChat}>
                <Unlock size={16} />
                Enable Chat
              </button>
            ) : (
              <>
                <div
                  className="my-color"
                  style={{ backgroundColor: visitorColor }}
                />
                <input
                  ref={inputRef}
                  type="text"
                  value={input}
                  onChange={(e) => setInput(e.target.value)}
                  onKeyDown={handleKeyDown}
                  placeholder="Type a message..."
                  maxLength={280}
                  disabled={!isConnected}
                />
                {recognitionRef.current && (
                  <button
                    className={`stt-button ${isListening ? 'listening' : ''}`}
                    onClick={toggleListening}
                    aria-label={isListening ? 'Stop listening' : 'Voice input'}
                  >
                    {isListening ? <MicOff size={16} /> : <Mic size={16} />}
                  </button>
                )}
                <button
                  className="send-button"
                  onClick={handleSend}
                  disabled={!input.trim() || !isConnected}
                  aria-label="Send message"
                >
                  <Send size={16} />
                </button>
              </>
            )}
          </div>
        </div>
      )}

      <style>{`
        .channel-toggle {
          position: fixed;
          z-index: 1000;
          width: 48px;
          height: 48px;
          border-radius: 50%;
          border: none;
          background: var(--color-bg-secondary);
          border: 2px solid var(--color-bg-tertiary);
          color: var(--color-text);
          cursor: pointer;
          display: flex;
          align-items: center;
          justify-content: center;
          transition: all 0.2s ease;
        }

        .channel-toggle:hover {
          border-color: var(--color-accent);
          transform: scale(1.05);
        }

        .text-toggle {
          bottom: 80px;
          right: 24px;
        }

        .message-badge {
          position: absolute;
          top: -4px;
          right: -4px;
          min-width: 18px;
          height: 18px;
          border-radius: 9px;
          background: #ef4444;
          color: white;
          font-size: 0.7rem;
          font-weight: 600;
          display: flex;
          align-items: center;
          justify-content: center;
          padding: 0 4px;
        }

        .text-channel {
          position: fixed;
          bottom: 80px;
          right: 24px;
          width: 340px;
          max-height: 450px;
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
          display: flex;
          flex-direction: column;
          z-index: 1000;
          box-shadow: 0 8px 32px rgba(0, 0, 0, 0.3);
          animation: slide-up 0.2s ease;
        }

        @keyframes slide-up {
          from {
            opacity: 0;
            transform: translateY(10px);
          }
          to {
            opacity: 1;
            transform: translateY(0);
          }
        }

        .channel-header {
          display: flex;
          align-items: center;
          gap: var(--space-sm);
          padding: var(--space-sm) var(--space-md);
          border-bottom: 1px solid var(--color-bg-tertiary);
        }

        .channel-header h3 {
          flex: 1;
          margin: 0;
          font-size: 0.9rem;
          color: var(--color-text);
        }

        .channel-status {
          display: flex;
          align-items: center;
          gap: 4px;
          font-size: 0.7rem;
          color: var(--color-text-muted);
        }

        .close-button {
          width: 24px;
          height: 24px;
          border-radius: 4px;
          border: none;
          background: transparent;
          color: var(--color-text-muted);
          cursor: pointer;
          display: flex;
          align-items: center;
          justify-content: center;
        }

        .close-button:hover {
          background: var(--color-bg-tertiary);
          color: var(--color-text);
        }

        .messages-container {
          flex: 1;
          overflow-y: auto;
          padding: var(--space-md);
          min-height: 200px;
          max-height: 300px;
        }

        .empty-state {
          display: flex;
          flex-direction: column;
          align-items: center;
          justify-content: center;
          height: 100%;
          color: var(--color-text-muted);
          text-align: center;
          gap: var(--space-xs);
        }

        .empty-state p {
          margin: 0;
        }

        .empty-hint {
          font-size: 0.8rem;
          opacity: 0.7;
        }

        .message {
          display: flex;
          align-items: flex-start;
          gap: var(--space-sm);
          padding: var(--space-xs) 0;
        }

        .message-avatar {
          width: 24px;
          height: 24px;
          border-radius: 50%;
          flex-shrink: 0;
        }

        .message-content {
          flex: 1;
          min-width: 0;
        }

        .message-sender {
          display: block;
          font-size: 0.75rem;
          font-weight: 600;
          margin-bottom: 2px;
        }

        .message-text {
          display: block;
          font-size: 0.85rem;
          color: var(--color-text);
          word-break: break-word;
        }

        .message-actions {
          display: flex;
          flex-direction: column;
          align-items: flex-end;
          gap: 2px;
        }

        .tts-button {
          width: 20px;
          height: 20px;
          border-radius: 4px;
          border: none;
          background: transparent;
          color: var(--color-text-muted);
          cursor: pointer;
          display: flex;
          align-items: center;
          justify-content: center;
          opacity: 0;
          transition: opacity 0.2s ease;
        }

        .message:hover .tts-button {
          opacity: 1;
        }

        .tts-button:hover {
          color: var(--color-accent);
        }

        .message-time {
          font-size: 0.65rem;
          color: var(--color-text-muted);
        }

        .input-container {
          display: flex;
          align-items: center;
          gap: var(--space-sm);
          padding: var(--space-sm) var(--space-md);
          border-top: 1px solid var(--color-bg-tertiary);
        }

        .enable-button {
          flex: 1;
          padding: var(--space-sm) var(--space-md);
          border-radius: 8px;
          border: 1px dashed var(--color-bg-tertiary);
          background: transparent;
          color: var(--color-text-muted);
          cursor: pointer;
          display: flex;
          align-items: center;
          justify-content: center;
          gap: var(--space-xs);
          font-size: 0.85rem;
          transition: all 0.2s ease;
        }

        .enable-button:hover {
          border-color: var(--color-accent);
          color: var(--color-accent);
        }

        .my-color {
          width: 24px;
          height: 24px;
          border-radius: 50%;
          flex-shrink: 0;
        }

        .input-container input {
          flex: 1;
          padding: var(--space-sm);
          border-radius: 8px;
          border: 1px solid var(--color-bg-tertiary);
          background: var(--color-bg);
          color: var(--color-text);
          font-size: 0.85rem;
        }

        .input-container input:focus {
          outline: none;
          border-color: var(--color-accent);
        }

        .stt-button {
          width: 32px;
          height: 32px;
          border-radius: 8px;
          border: none;
          background: var(--color-bg-tertiary);
          color: var(--color-text-muted);
          cursor: pointer;
          display: flex;
          align-items: center;
          justify-content: center;
          transition: all 0.2s ease;
        }

        .stt-button:hover {
          color: var(--color-text);
        }

        .stt-button.listening {
          background: #ef4444;
          color: white;
          animation: pulse-red 1s ease infinite;
        }

        @keyframes pulse-red {
          0%, 100% { box-shadow: 0 0 0 0 rgba(239, 68, 68, 0.4); }
          50% { box-shadow: 0 0 0 8px rgba(239, 68, 68, 0); }
        }

        .send-button {
          width: 32px;
          height: 32px;
          border-radius: 8px;
          border: none;
          background: var(--color-accent);
          color: white;
          cursor: pointer;
          display: flex;
          align-items: center;
          justify-content: center;
          transition: all 0.2s ease;
        }

        .send-button:hover:not(:disabled) {
          opacity: 0.9;
        }

        .send-button:disabled {
          opacity: 0.5;
          cursor: not-allowed;
        }

        @media (max-width: 480px) {
          .text-channel {
            right: 12px;
            left: 12px;
            width: auto;
            bottom: 72px;
          }

          .text-toggle {
            right: 12px;
            bottom: 68px;
          }
        }
      `}</style>
    </>
  );
});
