// @a0n/gnosis/forkjoin-app/client/components/stack-model.ts
export interface StackSegmentDefinition {
  label: string;
  ratio: number;
  color: string;
}

export type StackIconKey =
  | 'activity'
  | 'command'
  | 'compass'
  | 'database'
  | 'file-search'
  | 'git-branch'
  | 'globe'
  | 'hammer'
  | 'hard-drive'
  | 'key-round'
  | 'lock'
  | 'message-square'
  | 'monitor'
  | 'orbit'
  | 'radio'
  | 'refresh-cw'
  | 'settings'
  | 'shield'
  | 'sparkles'
  | 'wallet'
  | 'waves'
  | 'workflow'
  | 'brain'
  | 'zap';

export interface FoundationProductDefinition {
  id: string;
  name: string;
  layer: string;
  icon: StackIconKey;
  tagline: string;
  description: string;
  color: string;
  href: string;
  highlights: readonly string[];
  docsUrl?: string;
  sourceUrl?: string;
  install?: string;
  relatedProductIds?: readonly string[];
  segments?: readonly StackSegmentDefinition[];
}

function docs(slug: string): string {
  const normalizedSlug = slug
    .split('/')
    .filter((segment) => segment.length > 0)
    .map((segment) => encodeURIComponent(segment))
    .join('/');

  return `https://docs.aeonflux.dev/#/page/${normalizedSlug}`;
}

function repo(name: string): string {
  return `https://github.com/forkjoin-ai/${name}`;
}

function page(id: string): string {
  return `#/${id}`;
}

/**
 * Ordered from the user-facing surfaces at the top of the pyramid down to the
 * cryptographic base at the bottom.
 */
export const FOUNDATION_PRODUCTS_TOP_DOWN = [
  {
    id: 'memo',
    name: 'Memo',
    layer: 'Memory Apps',
    icon: 'message-square',
    tagline: 'A shared notebook that keeps ideas connected over time.',
    description:
      'Memo is the writing surface for ideas, meeting notes, and shared recall. It keeps notes connected to the wider memory layer instead of trapping them in isolated documents.',
    color: '#fb923c',
    href: page('memo'),
    docsUrl: docs('ebooks/119-app-memo-yoga/README'),
    highlights: [
      'shared notes that stay connected',
      'memory playback and linked context',
      'works alongside Todo and Fact',
    ],
    relatedProductIds: ['todo', 'fact', 'stargate', 'aeon-shell'],
  },
  {
    id: 'todo',
    name: 'Todo',
    layer: 'Memory Apps',
    icon: 'workflow',
    tagline: 'Nested task maps that stay local-first and easy to share.',
    description:
      'Todo gives teams and agents a place to break work into smaller pieces, keep progress in sync, and carry plans forward without losing context.',
    color: '#f59e0b',
    href: page('todo'),
    docsUrl: docs('ebooks/112-app-fractal-todo/README'),
    highlights: [
      'nested task graphs',
      'shared planning without a central task server',
      'identity handoff through card.yoga',
    ],
    relatedProductIds: ['memo', 'fact', 'iam', 'stargate'],
  },
  {
    id: 'fact',
    name: 'Fact',
    layer: 'Memory Apps',
    icon: 'file-search',
    tagline: 'A fact graph for what your team actually knows.',
    description:
      'Fact stores durable claims, sources, and relationships so people can search what is known, not just what was written down.',
    color: '#facc15',
    href: page('fact'),
    docsUrl: docs('ebooks/108-app-fractal-fact/README'),
    highlights: [
      'structured facts and linked sources',
      'fast graph lookup across shared knowledge',
      'pairs with Memo and Todo',
    ],
    relatedProductIds: ['memo', 'todo', 'stargate'],
  },
  {
    id: 'nexus',
    name: 'Nexus',
    layer: 'Groups',
    icon: 'orbit',
    tagline: 'The shared home for groups, rooms, and decisions.',
    description:
      'Nexus is where a phyle or team meets, talks, and coordinates. It brings shared presence, group decisions, and ongoing context into one place without turning the group into someone else’s platform.',
    color: '#c084fc',
    href: page('nexus'),
    highlights: [
      'group spaces and live coordination',
      'shared decision-making and presence',
      'works for people and agents together',
    ],
    relatedProductIds: ['zeitgeist', 'covenant', 'stargate'],
  },
  {
    id: 'zeitgeist',
    name: 'Zeitgeist',
    layer: 'Discovery',
    icon: 'sparkles',
    tagline:
      'Discovery that helps you notice what is happening across the network.',
    description:
      'Zeitgeist turns live activity into a discovery surface. It helps people notice emerging conversations, shared interests, and useful signals across the broader Aeon network.',
    color: '#a855f7',
    href: page('zeitgeist'),
    highlights: [
      'network discovery and activity signals',
      'shared context and trend awareness',
      'feeds groups, markets, and shell navigation',
    ],
    relatedProductIds: ['nexus', 'covenant', 'stargate'],
  },
  {
    id: 'covenant',
    name: 'Covenant Attention Market',
    layer: 'Markets',
    icon: 'wallet',
    tagline: 'A market layer for attention, incentives, and shared value.',
    description:
      'Covenant is where phyles can turn coordination into offers, rewards, and reciprocal value flows. In today’s stack it is a public surface and routing contract, with the deeper market runtime still to come.',
    color: '#8b5cf6',
    href: page('covenant'),
    docsUrl: docs(
      'ebooks/126-covenant-attention-marketplace-and-reciprocal-value-flows/README'
    ),
    highlights: [
      'attention and incentive routing',
      'market entry tied to phyle identity',
      'works through Stargate, Nexus, and discovery surfaces',
    ],
    relatedProductIds: ['nexus', 'zeitgeist', 'app-store', 'stargate'],
  },
  {
    id: 'app-store',
    name: 'App Store',
    layer: 'Distribution',
    icon: 'command',
    tagline: 'A trusted catalog for apps, agents, and installable tools.',
    description:
      'The App Store helps people discover what can be added to Aeon Shell, from full apps and agents to smaller capabilities that extend how the system works.',
    color: '#7c3aed',
    href: page('app-store'),
    highlights: [
      'discover shell-ready apps and agents',
      'install capabilities without hunting for files',
      'connects builders with end users',
    ],
    relatedProductIds: ['stargate', 'aeon-shell', 'memo', 'todo', 'fact'],
  },
  {
    id: 'stargate',
    name: 'Stargate',
    layer: 'Portals',
    icon: 'compass',
    tagline: 'The entry point that helps people discover a phyle and join it.',
    description:
      'Stargate is the public entry layer for phyles. It introduces a group, shows what it stands for, checks what is needed to join, and hands people into the right Aeon surface once trust is established.',
    color: '#6366f1',
    href: page('stargate'),
    docsUrl: docs('ebooks/29-stargate-mutual-delegation-virtual-phyles/README'),
    highlights: [
      'discover phyles and review their mission',
      'join through a mutual trust handshake',
      'opens Nexus, Zeitgeist, Covenant, and the App Store',
    ],
    relatedProductIds: [
      'nexus',
      'zeitgeist',
      'covenant',
      'app-store',
      'aeon-shell',
    ],
    segments: [
      { label: 'Nexus', ratio: 0.25, color: '#c084fc' },
      { label: 'Zeitgeist', ratio: 0.25, color: '#a855f7' },
      { label: 'Covenant', ratio: 0.25, color: '#8b5cf6' },
      { label: 'App Store', ratio: 0.25, color: '#7c3aed' },
    ],
  },
  {
    id: 'aeon-shell',
    name: 'Aeon Shell',
    layer: 'Operating Surface',
    icon: 'monitor',
    tagline: 'The everyday operating surface for the Aeon world.',
    description:
      'Aeon Shell is the local-first environment that opens apps, routes people into shared spaces, and keeps aeon:// links, launchers, maps, and workspace views tied together.',
    color: '#3b82f6',
    href: page('aeon-shell'),
    docsUrl: docs(
      'ebooks/26-aeon-shell-building-terminal-user-interfaces-tui/README'
    ),
    sourceUrl: repo('aeon-shell'),
    highlights: [
      'local-first shell and launcher',
      'opens routes, maps, and storefronts',
      'ties Stargate and app surfaces together',
    ],
    relatedProductIds: ['stargate', 'app-store', 'flux'],
  },
  {
    id: 'flux',
    name: 'Aeon-Flux',
    layer: 'Frontend',
    icon: 'waves',
    tagline: 'The frontend framework for collaborative, edge-native surfaces.',
    description:
      'Aeon-Flux is the UI framework behind collaborative, local-first apps across the stack, with rendering patterns built for shared state and edge delivery.',
    color: '#06b6d4',
    href: page('flux'),
    docsUrl: docs('packages/aeon-flux/index'),
    sourceUrl: repo('aeon-flux'),
    install: 'npx create-aeon-flux my-app',
    highlights: [
      'edge-first page rendering',
      'collaborative runtime primitives',
      'the UI layer used across the ecosystem',
    ],
    relatedProductIds: ['aeon-shell', 'forge', 'goodchild', 'preferences'],
  },
  {
    id: 'forge',
    name: 'Aeon Forge',
    layer: 'Development',
    icon: 'hammer',
    tagline: 'Developer tools for shaping and shipping on the stack.',
    description:
      'Forge gives builders the commands, planning tools, and operator surfaces needed to create Aeon apps and get them ready to publish.',
    color: '#14b8a6',
    href: page('forge'),
    docsUrl: docs('ebooks/16-deploying-with-aeon-forge/README'),
    sourceUrl: repo('aeon-forge'),
    highlights: [
      'build and deployment planning',
      'CLI and operator tools',
      'supports Flux, Forgo, and shell delivery',
    ],
    relatedProductIds: ['flux', 'forgo', 'forgo-logs'],
  },
  {
    id: 'forgo',
    name: 'Forgo',
    layer: 'Deployment',
    icon: 'git-branch',
    tagline: 'Edge deployment without the usual platform sprawl.',
    description:
      'Forgo handles publish flows, promotions, and automation for Aeon apps so teams can ship to the edge without carrying a large operations footprint.',
    color: '#10b981',
    href: page('forgo'),
    docsUrl: docs('products/forgo'),
    highlights: [
      'publish and promotion flows',
      'event-driven deployment automation',
      'delivery layer behind Forge',
    ],
    relatedProductIds: ['forge', 'forgo-logs', 'flux'],
  },
  {
    id: 'forgo-logs',
    name: 'OTel Logs',
    layer: 'Observability',
    icon: 'activity',
    tagline: 'See what happened across build, deploy, and runtime.',
    description:
      'OpenTelemetry-based traces and logs give teams one place to inspect how apps behave from rollout to live operation.',
    color: '#22c55e',
    href: page('forgo-logs'),
    docsUrl: docs('products/forgo-logs'),
    highlights: [
      'OpenTelemetry traces and logs',
      'runtime health and quota visibility',
      'the monitoring layer for Forgo and live apps',
    ],
    relatedProductIds: ['forge', 'forgo', 'edgework'],
  },
  {
    id: 'goodchild',
    name: 'Goodchild',
    layer: 'Entitlements',
    icon: 'key-round',
    tagline: 'Feature access controlled by keys you hold.',
    description:
      'Goodchild lets apps turn features on or off using local cryptographic grants instead of a remote flag service.',
    color: '#84cc16',
    href: page('goodchild'),
    docsUrl: docs('ebooks/79-aeon-flags-feature-management/README'),
    sourceUrl: repo('aeon-flags'),
    install: 'bun add @a0n/aeon-flags',
    highlights: [
      'UCAN-backed feature checks',
      'no polling SaaS control plane',
      'predictable routing and access rules',
    ],
    relatedProductIds: ['preferences', 'flux', 'iam', 'aegis'],
  },
  {
    id: 'preferences',
    name: 'Aeon Preferences',
    layer: 'User State',
    icon: 'settings',
    tagline: 'Your settings, saved locally and shared where they need to be.',
    description:
      'Aeon Preferences manages personal settings, scoped namespaces, and protected vault data so the rest of the ecosystem can respect how each person wants things to work.',
    color: '#eab308',
    href: page('preferences'),
    docsUrl: docs('packages/aeon-preferences/index'),
    sourceUrl: repo('aeon-preferences'),
    install: 'bun add @a0n/aeon-preferences',
    highlights: [
      'local-first personal settings',
      'protected vault space for sensitive material',
      'combines personal state with entitlements',
    ],
    relatedProductIds: ['goodchild', 'dashrelay', 'iam', 'aegis'],
  },
  {
    id: 'neural',
    name: 'Neural',
    layer: 'Model Runtime',
    icon: 'brain',
    tagline: 'Run intelligence close to the person using it.',
    description:
      'Neural brings model execution onto the device and the edge so apps can respond quickly, protect privacy, and avoid paid AI lock-in.',
    color: '#38bdf8',
    href: page('neural'),
    docsUrl: docs('packages/neural/index'),
    highlights: [
      'browser-native inference',
      'distributed learning and collective models',
      'feeds semantic reasoning into deeper stack layers',
    ],
    relatedProductIds: ['edgework', 'flux', 'dash', 'aeon'],
  },
  {
    id: 'edgework',
    name: 'Edgework',
    layer: 'Semantic Compute',
    icon: 'globe',
    tagline: 'Compute services that stay close to the user and the data.',
    description:
      'Edgework handles the heavier work around inference, tool execution, and gateway logic so Aeon apps can do useful work without shipping everything back to a central backend.',
    color: '#60a5fa',
    href: page('edgework'),
    docsUrl: docs('products/edgework'),
    highlights: [
      'compute that can run from browser to edge to cloud',
      'agent and MCP execution surfaces',
      'useful automation inside the runtime',
    ],
    relatedProductIds: ['neural', 'dashrelay', 'forgo-logs'],
  },
  {
    id: 'dashrelay',
    name: 'DashRelay',
    layer: 'Trust Log',
    icon: 'zap',
    tagline: 'Shared sync, transport, and presence in one fabric.',
    description:
      'DashRelay combines storage updates and live relay transport so local-first apps can stay in sync across devices and collaborators.',
    color: '#34d399',
    href: page('dashrelay'),
    docsUrl: docs('products/dashrelay'),
    highlights: [
      'encrypted updates between devices',
      'presence and realtime synchronization',
      'the bridge between storage and transport',
    ],
    relatedProductIds: ['relay', 'dash', 'preferences', 'edgework'],
    segments: [
      { label: 'Dash', ratio: 0.5, color: '#f59e0b' },
      { label: 'Relay', ratio: 0.5, color: '#2dd4bf' },
    ],
  },
  {
    id: 'relay',
    name: 'Relay',
    layer: 'Transport',
    icon: 'radio',
    tagline: 'Realtime communication for local-first collaboration.',
    description:
      'Relay carries presence, streams, and encrypted messages between people, devices, and agents that need to stay connected.',
    color: '#2dd4bf',
    href: page('relay'),
    highlights: [
      'presence and streaming transport',
      'encrypted realtime communication',
      'pairs with Dash and Aeon underneath local-first apps',
    ],
    relatedProductIds: ['dashrelay', 'dash', 'aeon'],
  },
  {
    id: 'dash',
    name: 'Dash',
    layer: 'Data',
    icon: 'hard-drive',
    tagline: 'Storage built for local-first apps and AI-friendly data.',
    description:
      'Dash stores app state close to the user while still making it practical to search, sync, and reuse that data across the stack.',
    color: '#f59e0b',
    href: page('dash'),
    docsUrl: docs('packages/dash/index'),
    highlights: [
      'local-first data tiers',
      'AI-native storage and retrieval',
      'the durable state substrate beneath DashRelay',
    ],
    relatedProductIds: ['relay', 'dashrelay', 'aeon', 'preferences'],
  },
  {
    id: 'aeon',
    name: 'Aeon',
    layer: 'Sync',
    icon: 'refresh-cw',
    tagline: 'The sync layer that keeps shared state from drifting.',
    description:
      'Aeon provides the sync rules, schema evolution, and distributed state primitives that let apps collaborate without a central source of truth.',
    color: '#5599cc',
    href: page('aeon'),
    docsUrl: docs('packages/aeon/index'),
    highlights: [
      'distributed state synchronization',
      'schema evolution and versioning',
      'shared sync contract beneath Dash and Relay',
    ],
    relatedProductIds: ['dash', 'relay', 'iam', 'aegis'],
  },
  {
    id: 'iam',
    name: 'Aeon IAM',
    layer: 'Identity Hub',
    icon: 'lock',
    tagline: 'Identity handoff rooted in your device, not a platform account.',
    description:
      'Aeon IAM connects sign-in providers, starter grants, and identity flows so apps and shell surfaces can recognize the same person without copying their data everywhere.',
    color: '#f472b6',
    href: page('iam'),
    sourceUrl: repo('aeon-iam'),
    highlights: [
      'device-rooted UCAN hub',
      'card.yoga-style identity handoff surfaces',
      'starter grants for apps, shell surfaces, and agents',
    ],
    relatedProductIds: ['aegis', 'goodchild', 'preferences', 'todo'],
  },
  {
    id: 'aegis',
    name: 'Aegis',
    layer: 'Crypto + UCAN',
    icon: 'shield',
    tagline: 'The trust layer for keys, proofs, and capability-based access.',
    description:
      'Aegis is the cryptographic foundation under the stack. It manages keys, UCAN grants, and privacy-preserving proofs so higher layers can trust what they are opening.',
    color: '#ec4899',
    href: page('aegis'),
    docsUrl: docs('packages/zk-encryption/index'),
    sourceUrl: repo('aegis'),
    highlights: [
      'UCAN permission chains',
      'privacy-preserving proofs',
      'vault-backed crypto building blocks',
    ],
    relatedProductIds: ['iam', 'goodchild', 'preferences', 'aeon'],
    segments: [
      { label: 'UCAN', ratio: 0.34, color: '#f472b6' },
      { label: 'ZK', ratio: 0.33, color: '#fb7185' },
      { label: 'Vault', ratio: 0.33, color: '#ec4899' },
    ],
  },
] as const;

export type FoundationProductId =
  (typeof FOUNDATION_PRODUCTS_TOP_DOWN)[number]['id'];

export interface FoundationProductRecord extends FoundationProductDefinition {
  id: FoundationProductId;
  relatedProductIds?: readonly FoundationProductId[];
}

export interface StackLayerDefinition {
  id: string;
  name: string;
  layer: string;
  icon: StackIconKey;
  description: string;
  color: string;
  href: string;
  segments?: readonly StackSegmentDefinition[];
}

export const STACK_LAYERS_TOP_DOWN: readonly StackLayerDefinition[] = [
  {
    id: 'experience-layer',
    name: 'Experience',
    layer: 'Apps',
    icon: 'monitor',
    description:
      'People-facing surfaces where work happens across apps, agents, and domain workflows.',
    color: '#fb923c',
    href: page('memo'),
  },
  {
    id: 'collective-layer',
    name: 'Collective',
    layer: 'Groups + Discovery',
    icon: 'orbit',
    description:
      'Shared spaces for groups, discovery, and reciprocal coordination across the network.',
    color: '#c084fc',
    href: page('nexus'),
    segments: [
      { label: 'Nexus', ratio: 0.34, color: '#c084fc' },
      { label: 'Zeitgeist', ratio: 0.33, color: '#a855f7' },
      { label: 'Covenant', ratio: 0.33, color: '#8b5cf6' },
    ],
  },
  {
    id: 'portal-layer',
    name: 'Portals',
    layer: 'Access + Distribution',
    icon: 'compass',
    description:
      'Entry points and catalog surfaces that route people into the right context and tools.',
    color: '#7c3aed',
    href: page('stargate'),
    segments: [
      { label: 'Stargate', ratio: 0.5, color: '#6366f1' },
      { label: 'App Store', ratio: 0.5, color: '#7c3aed' },
    ],
  },
  {
    id: 'surface-layer',
    name: 'Operating Surface',
    layer: 'Shell + UI Primitives',
    icon: 'monitor',
    description:
      'The shell and interface primitives that unify the ecosystem, including shared components like aeon-icons.',
    color: '#3b82f6',
    href: page('aeon-shell'),
  },
  {
    id: 'builder-layer',
    name: 'Builder',
    layer: 'Framework + Delivery',
    icon: 'hammer',
    description:
      'Build and ship toolchain for the ecosystem, spanning aeon-flux, forge/forgo, and aeon-pipelines.',
    color: '#14b8a6',
    href: page('flux'),
    segments: [
      { label: 'Aeon-Flux', ratio: 0.34, color: '#06b6d4' },
      { label: 'Forge/Forgo', ratio: 0.33, color: '#14b8a6' },
      { label: 'Pipelines', ratio: 0.33, color: '#10b981' },
    ],
  },
  {
    id: 'intelligence-layer',
    name: 'Intelligence',
    layer: 'Model + Semantic Compute',
    icon: 'brain',
    description:
      'Local and edge intelligence runtime for inference, semantic reasoning, and automation.',
    color: '#38bdf8',
    href: page('neural'),
    segments: [
      { label: 'Neural', ratio: 0.5, color: '#38bdf8' },
      { label: 'Edgework', ratio: 0.5, color: '#60a5fa' },
    ],
  },
  {
    id: 'memory-layer',
    name: 'Memory',
    layer: 'Knowledge Layer',
    icon: 'message-square',
    description:
      'Shared durable context for notes, tasks, and facts that persists across teams, agents, and time.',
    color: '#f59e0b',
    href: page('memo'),
    segments: [
      { label: 'Memo', ratio: 0.34, color: '#fb923c' },
      { label: 'Todo', ratio: 0.33, color: '#f59e0b' },
      { label: 'Fact', ratio: 0.33, color: '#facc15' },
    ],
  },
  {
    id: 'datastore-layer',
    name: 'Datastore',
    layer: 'State + Preferences',
    icon: 'database',
    description:
      'Local-first storage and state fabric, from structured app data to user preferences and aeon-everything.',
    color: '#60a5fa',
    href: page('dash'),
    segments: [
      { label: 'Dash', ratio: 0.5, color: '#f59e0b' },
      { label: 'Preferences', ratio: 0.3, color: '#eab308' },
      { label: 'Everything', ratio: 0.2, color: '#38bdf8' },
    ],
  },
  {
    id: 'sync-layer',
    name: 'Sync Fabric',
    layer: 'Convergence + Presence',
    icon: 'refresh-cw',
    description:
      'Replication and trust-log coordination that keeps distributed state convergent and collaborative.',
    color: '#34d399',
    href: page('dashrelay'),
    segments: [
      { label: 'DashRelay', ratio: 0.6, color: '#34d399' },
      { label: 'Aeon', ratio: 0.4, color: '#5599cc' },
    ],
  },
  {
    id: 'transport-layer',
    name: 'Transport',
    layer: 'Realtime Network',
    icon: 'radio',
    description:
      'Encrypted messaging and streaming transport between users, devices, and agents.',
    color: '#2dd4bf',
    href: page('relay'),
  },
  {
    id: 'identity-layer',
    name: 'Identity',
    layer: 'IAM + Entitlements',
    icon: 'lock',
    description:
      'Identity handoff and capability checks that determine who can access what and when.',
    color: '#f472b6',
    href: page('iam'),
    segments: [
      { label: 'IAM', ratio: 0.6, color: '#f472b6' },
      { label: 'Goodchild', ratio: 0.4, color: '#84cc16' },
    ],
  },
  {
    id: 'trust-layer',
    name: 'Trust Core',
    layer: 'Crypto + UCAN',
    icon: 'shield',
    description:
      'Cryptographic root of trust for keys, delegation chains, proofs, and secure vault operations.',
    color: '#ec4899',
    href: page('aegis'),
    segments: [
      { label: 'UCAN', ratio: 0.34, color: '#f472b6' },
      { label: 'ZK', ratio: 0.33, color: '#fb7185' },
      { label: 'Vault', ratio: 0.33, color: '#ec4899' },
    ],
  },
] as const;

export const FOUNDATION_PRODUCTS_BY_ID = FOUNDATION_PRODUCTS_TOP_DOWN.reduce(
  (registry, product) => {
    registry[product.id] = product;
    return registry;
  },
  {} as Record<FoundationProductId, FoundationProductRecord>
);

/**
 * Handles the gnosis get Foundation Product workflow.
 */
export function getFoundationProduct(
  productId: FoundationProductId
): FoundationProductRecord {
  return FOUNDATION_PRODUCTS_BY_ID[productId];
}

/**
 * Handles the gnosis find Foundation Product By Hash workflow.
 */
export function findFoundationProductByHash(
  hash: string
): FoundationProductRecord | undefined {
  return FOUNDATION_PRODUCTS_TOP_DOWN.find((product) => product.href === hash);
}

/**
 * Handles the gnosis get Foundation Stack Neighbors workflow.
 */
export function getFoundationStackNeighbors(productId: FoundationProductId): {
  above?: FoundationProductRecord;
  below?: FoundationProductRecord;
} {
  const index = FOUNDATION_PRODUCTS_TOP_DOWN.findIndex(
    (product) => product.id === productId
  );

  if (index === -1) {
    return {};
  }

  return {
    above: index > 0 ? FOUNDATION_PRODUCTS_TOP_DOWN[index - 1] : undefined,
    below:
      index < FOUNDATION_PRODUCTS_TOP_DOWN.length - 1
        ? FOUNDATION_PRODUCTS_TOP_DOWN[index + 1]
        : undefined,
  };
}
