// @a0n/gnosis/scripts/convert-formal-ledger.ts
/**
 * convert-formal-ledger.ts -- Batch convert TLA+ and Lean to .gg proof topologies
 *
 * Reads every .tla and .lean file in the formal ledger and emits
 * one .gg proof topology per file into gnosis/examples/proofs/.
 *
 * Usage:
 *   gnode run scripts/convert-formal-ledger.ts
 *
 * The entire paper -- 110 TLA+ specs, 155 Lean theorem files --
 * becomes 265 .gg proof topologies. One file per theorem.
 * The topology formalizes the proof. The clunky Lean/TLA+ is just the compilation target.
 */

import {
  readFileSync,
  writeFileSync,
  mkdirSync,
  readdirSync,
  existsSync,
} from 'fs';
import { join, basename, extname } from 'path';
import { tlaToGg, leanToGg } from '../../aeon-logic/src/buleyean-proof';

const TLA_DIR = join(__dirname, '../tla');
const LEAN_DIR = join(__dirname, '../lean/Lean/ForkRaceFoldTheorems');
const OUTPUT_DIR = join(__dirname, '../examples/proofs');
const REYNOLDS_BFT_PRIVATE_PUBLIC_THEOREMS = new Set([
  'quorumSafe_of_chunks_ge_stages',
  'majoritySafe_of_chunks_ge_stages',
  'majoritySafe_of_two_chunks_gt_stages',
  'quorumSafe_of_three_chunks_gt_two_stages',
  'mergeAll_is_quorumSafe',
  'quorumFold_is_majoritySafe',
  'syncRequired_not_majoritySafe',
]);

function ensureDir(dir: string): void {
  if (!existsSync(dir)) {
    mkdirSync(dir, { recursive: true });
  }
}

function sanitizeGgId(raw: string): string {
  return (
    raw
      .replace(/[^A-Za-z0-9_]/g, '_')
      .replace(/_+/g, '_')
      .replace(/^_+/, '')
      .replace(/_+$/, '')
      .toLowerCase() || 'unnamed'
  );
}

function stripReynoldsHelperTheorems(gg: string): string {
  const removedIds = new Set(
    [...REYNOLDS_BFT_PRIVATE_PUBLIC_THEOREMS].map((name) => sanitizeGgId(name))
  );
  const lines = gg.split('\n');
  const keptTheoremIds: string[] = [];
  const keptProvedIds: string[] = [];
  const filteredLines: string[] = [];

  for (const line of lines) {
    const theoremDecl = line.match(
      /^\(([^:]+): Lemma { bules: '1', proposition: '([^']+)' }\)$/
    );
    if (theoremDecl) {
      const theoremId = theoremDecl[1];
      if (removedIds.has(theoremId)) {
        continue;
      }
      keptTheoremIds.push(theoremId);
      filteredLines.push(line);
      continue;
    }

    const provedDecl = line.match(/^\(([^:]+_proved): Lemma { bules: '0' }\)$/);
    if (provedDecl) {
      const theoremId = provedDecl[1].replace(/_proved$/, '');
      if (removedIds.has(theoremId)) {
        continue;
      }
      keptProvedIds.push(provedDecl[1]);
      filteredLines.push(line);
      continue;
    }

    const rejectEdge = line.match(/^\(([^)]+)\)-\[:REJECT /);
    if (rejectEdge && removedIds.has(rejectEdge[1])) {
      continue;
    }

    filteredLines.push(line);
  }

  return filteredLines
    .map((line) => {
      if (line.startsWith('(theorem: Theorem {')) {
        return line.replace(
          /bules: '\d+'/u,
          `bules: '${keptTheoremIds.length}'`
        );
      }

      if (line.startsWith('(theorem)-[:FORK')) {
        return `(theorem)-[:FORK { reason: 'prove each theorem independently' }]->(${keptTheoremIds.join(
          ' | '
        )})`;
      }

      if (line.includes("[:FOLD { reason: 'all theorems proved' }]")) {
        return `(${keptProvedIds.join(
          ' | '
        )})-[:FOLD { reason: 'all theorems proved' }]->(qed: Theorem { bules: '0', proposition: 'ForkRaceFoldTheorems QED' })`;
      }

      return line;
    })
    .join('\n');
}

function convertTlaFiles(): number {
  if (!existsSync(TLA_DIR)) {
    console.log(`TLA+ directory not found: ${TLA_DIR}`);
    return 0;
  }

  const tlaFiles = readdirSync(TLA_DIR).filter(
    (f) => f.endsWith('.tla') && !f.startsWith('__tmp')
  );

  let converted = 0;
  for (const file of tlaFiles) {
    const name = basename(file, '.tla');
    const source = readFileSync(join(TLA_DIR, file), 'utf-8');
    const gg = tlaToGg(source);
    const outPath = join(OUTPUT_DIR, `${name}.gg`);
    writeFileSync(outPath, gg);
    converted++;
    console.log(`  TLA+ → GG: ${name}.tla → ${name}.gg`);
  }

  return converted;
}

function convertLeanFiles(): number {
  if (!existsSync(LEAN_DIR)) {
    console.log(`Lean directory not found: ${LEAN_DIR}`);
    return 0;
  }

  const leanFiles = readdirSync(LEAN_DIR).filter((f) => f.endsWith('.lean'));

  let converted = 0;
  for (const file of leanFiles) {
    const name = basename(file, '.lean');
    const source = readFileSync(join(LEAN_DIR, file), 'utf-8');
    const gg =
      name === 'ReynoldsBFT'
        ? stripReynoldsHelperTheorems(leanToGg(source))
        : leanToGg(source);
    const outPath = join(OUTPUT_DIR, `${name}.gg`);
    const tlaPath = join(TLA_DIR, `${name}.tla`);

    // Don't overwrite TLA+-generated files -- Lean may have more detail.
    // Only append '-lean' when a matching TLA+ source actually exists.
    const finalPath = existsSync(tlaPath)
      ? join(OUTPUT_DIR, `${name}-lean.gg`)
      : outPath;

    writeFileSync(finalPath, gg);
    converted++;
    console.log(`  Lean → GG: ${name}.lean → ${basename(finalPath)}`);
  }

  return converted;
}

/**
 * Runs the gnosis command-line workflow.
 */
export function main(): void {
  console.log('Converting formal ledger to .gg proof topologies...');
  console.log(`Output: ${OUTPUT_DIR}`);
  console.log('');

  ensureDir(OUTPUT_DIR);

  console.log('=== TLA+ Specifications ===');
  const tlaCount = convertTlaFiles();
  console.log(`\n  ${tlaCount} TLA+ specs converted`);

  console.log('\n=== Lean 4 Theorems ===');
  const leanCount = convertLeanFiles();
  console.log(`\n  ${leanCount} Lean theorems converted`);

  console.log(`\n=== Total: ${tlaCount + leanCount} .gg proof topologies ===`);
  console.log('The entire paper is now topology.');
}
