// @a0n/gnosis/scripts/generate-imdb-top-250-monomyths.ts
/**
 * generate-imdb-top-250-monomyths.ts -- Generate a GG monomyth corpus from the IMDb Top 250 snapshot
 *
 * Source surface:
 *   https://www.listchallenges.com/imdb2026
 *
 * That page is a static snapshot of the IMDb Top 250 at 2026-01-01 and is
 * accessible to automation when IMDb's own chart is WAF-blocked.
 *
 * Usage:
 *   pnpm run gnode -- run open-source/gnosis/scripts/generate-imdb-top-250-monomyths.ts --export main
 */

import { mkdirSync, readdirSync, rmSync, writeFileSync } from 'fs';
import { request as httpRequest } from 'http';
import { request as httpsRequest } from 'https';
import { dirname, join } from 'path';
import { fileURLToPath } from 'url';
import { decode } from 'html-entities';

interface RankedFilm {
  rank: number;
  title: string;
  releaseYear?: string;
  page: number;
  sourceUrl: string;
}

const SNAPSHOT_DATE = '2026-01-01';
const BASE_URL = 'https://www.listchallenges.com/imdb2026';
const SCRIPT_DIR = dirname(fileURLToPath(import.meta.url));
const OUTPUT_DIR = join(
  SCRIPT_DIR,
  '../examples/proofs/imdb-top-250-monomyths'
);
const TEST_FILE = join(OUTPUT_DIR, 'imdb-top-250-monomyths.test.gg');

function getPageUrl(page: number): string {
  return page === 1 ? BASE_URL : `${BASE_URL}/list/${page}`;
}

function padRank(rank: number): string {
  return String(rank).padStart(3, '0');
}

function stripTags(value: string): string {
  return decode(value.replace(/<[^>]+>/g, ' '))
    .replace(/\s+/g, ' ')
    .trim();
}

function slugify(value: string): string {
  const normalized = value
    .normalize('NFKD')
    .replace(/[\u0300-\u036f]/g, '')
    .replace(/&/g, ' and ')
    .replace(/['’]/g, '');

  return (
    normalized
      .replace(/[^A-Za-z0-9]+/g, '-')
      .replace(/^-+|-+$/g, '')
      .replace(/-+/g, '-')
      .toLowerCase() || 'untitled'
  );
}

function ggQuote(value: string): string {
  if (value.includes('"')) {
    throw new Error(`GG double-quoted properties cannot encode ${value}`);
  }

  return `"${value}"`;
}

function ggSingleQuote(value: string): string {
  if (value.includes("'")) {
    throw new Error(`GG single-quoted properties cannot encode ${value}`);
  }

  return `'${value}'`;
}

function parseTitle(
  rawTitle: string
): Pick<RankedFilm, 'title' | 'releaseYear'> {
  const match = rawTitle.match(/^(.*?)(?: \((\d{4})\))?$/u);
  if (!match) {
    return { title: rawTitle };
  }

  const title = match[1]?.trim() || rawTitle;
  const releaseYear = match[2]?.trim();
  return releaseYear ? { title, releaseYear } : { title };
}

function downloadText(url: string): Promise<string> {
  return new Promise((resolve, reject) => {
    const transport = url.startsWith('https:') ? httpsRequest : httpRequest;
    const req = transport(
      url,
      {
        headers: {
          'user-agent':
            'Mozilla/5.0 (compatible; codex/1.0; +https://openai.com)',
        },
      },
      (res) => {
        const statusCode = res.statusCode ?? 0;
        const location = res.headers.location;

        if (location && [301, 302, 303, 307, 308].includes(statusCode)) {
          const redirected = new URL(location, url).toString();
          res.resume();
          resolve(downloadText(redirected));
          return;
        }

        if (statusCode < 200 || statusCode >= 300) {
          res.resume();
          reject(new Error(`Request failed for ${url}: ${statusCode}`));
          return;
        }

        const chunks: string[] = [];
        res.setEncoding('utf8');
        res.on('data', (chunk: string) => chunks.push(chunk));
        res.on('end', () => resolve(chunks.join('')));
      }
    );

    req.on('error', reject);
    req.end();
  });
}

function parsePage(
  html: string,
  page: number,
  sourceUrl: string
): RankedFilm[] {
  const itemRegex =
    /<div class="item-name">([\s\S]*?)<\/div>\s*<div class="item-rank">(\d+)<\/div>/g;
  const entries: RankedFilm[] = [];

  let match: RegExpExecArray | null;
  while ((match = itemRegex.exec(html)) !== null) {
    const rawTitle = stripTags(match[1] ?? '');
    const rank = Number.parseInt(match[2] ?? '', 10);
    if (!rawTitle || !Number.isFinite(rank)) {
      continue;
    }

    entries.push({
      rank,
      page,
      sourceUrl,
      ...parseTitle(rawTitle),
    });
  }

  return entries;
}

function validateEntries(entries: RankedFilm[]): void {
  if (entries.length !== 250) {
    throw new Error(`Expected 250 ranked films, found ${entries.length}`);
  }

  const seenRanks = new Set<number>();
  for (const entry of entries) {
    if (seenRanks.has(entry.rank)) {
      throw new Error(`Duplicate rank ${entry.rank}`);
    }
    seenRanks.add(entry.rank);
  }

  for (let rank = 1; rank <= 250; rank += 1) {
    if (!seenRanks.has(rank)) {
      throw new Error(`Missing rank ${rank}`);
    }
  }
}

function getFileName(entry: RankedFilm): string {
  return `${padRank(entry.rank)}-${slugify(entry.title)}.gg`;
}

function getTheoremId(entry: RankedFilm): string {
  return `IMDB_TOP_250_MONOMYTH_${padRank(entry.rank)}_${slugify(entry.title)
    .replace(/-/g, '_')
    .toUpperCase()}`;
}

function buildModule(entry: RankedFilm): string {
  const theoremId = getTheoremId(entry);
  const filmFile = getFileName(entry);
  const props = [
    `title: ${ggQuote(entry.title)}`,
    `rank: ${ggQuote(String(entry.rank))}`,
    `canon: ${ggQuote('imdb-top-250')}`,
    `snapshot_date: ${ggQuote(SNAPSHOT_DATE)}`,
    `source: ${ggQuote(entry.sourceUrl)}`,
    `source_page: ${ggQuote(String(entry.page))}`,
    `model: ${ggQuote('broad-monomyth')}`,
  ];

  if (entry.releaseYear) {
    props.splice(2, 0, `release_year: ${ggQuote(entry.releaseYear)}`);
  }

  return [
    `// ${filmFile} -- rank ${entry.rank} in the IMDb Top 250 static snapshot at ${SNAPSHOT_DATE}.`,
    '// Broad monomyth witness, not a screenplay-close formalization.',
    '',
    `(witness: Theorem { theorem: ${ggQuote(theoremId)}, status: ${ggQuote(
      'documentation-layer'
    )} })`,
    `(film: StoryWitness { ${props.join(', ')} })`,
    `(home: SemioticChannel { state: ${ggQuote(
      'ordinary-world'
    )}, deficit: ${ggQuote('positive')} })`,
    `(call: ForkEvent { state: ${ggQuote(
      'call-to-adventure'
    )}, live_paths: ${ggQuote('3')} })`,
    `(refusal: Branch { state: ${ggQuote('refusal-trace')}, weight: ${ggQuote(
      'nonzero'
    )} })`,
    `(answer: Branch { state: ${ggQuote('answer-the-call')}, weight: ${ggQuote(
      'positive'
    )} })`,
    `(mentor: SharedContext { state: ${ggQuote(
      'mentor-aid'
    )}, effect: ${ggQuote('reduce-effective-deficit')} })`,
    `(threshold: Threshold { state: ${ggQuote('crossing-the-threshold')} })`,
    `(trial_allies: Trial { state: ${ggQuote('allies-and-enemies')} })`,
    `(trial_temptation: Trial { state: ${ggQuote(
      'temptation-and-false-boon'
    )} })`,
    `(trial_abyss: Trial { state: ${ggQuote('approach-to-abyss')} })`,
    `(void_boundary: VoidBoundary { state: ${ggQuote(
      'rejection-memory'
    )}, monotone: ${ggQuote('true')} })`,
    `(ordeal: DeterministicCollapse { state: ${ggQuote(
      'ordeal'
    )}, paid: ${ggQuote('true')} })`,
    `(boon: Settlement { state: ${ggQuote('reward')}, deficit: ${ggQuote(
      'reduced'
    )} })`,
    `(reentry_arc: EmpathyTransform { state: ${ggQuote(
      'reentry-with-elixir'
    )}, exports: ${ggQuote('context')} })`,
    `(result: Result { claim: ${ggQuote('coarse-imdb-monomyth-witness')} })`,
    '',
    '(witness)-[:PROCESS]->(home)',
    `(home)-[:PROCESS { reason: ${ggQuote(
      'popular-culture arcs begin inside a legible ordinary world'
    )} }]->(call)`,
    `(call)-[:FORK { reason: ${ggQuote(
      'the call exposes refusal, answer, and aid as simultaneous narrative pressure'
    )} }]->(refusal | answer | mentor)`,
    `(refusal | answer | mentor)-[:FOLD { strategy: ${ggQuote(
      'context-assisted-threshold-crossing'
    )} }]->(threshold)`,
    `(threshold)-[:FORK { reason: ${ggQuote(
      'crossing the threshold opens a trial field rather than a single move'
    )} }]->(trial_allies | trial_temptation | trial_abyss)`,
    `(trial_allies | trial_temptation | trial_abyss)-[:FOLD { strategy: ${ggQuote(
      'accumulate-void-boundary'
    )} }]->(void_boundary)`,
    `(void_boundary)-[:PROCESS { reason: ${ggQuote(
      'rejection history sharpens the path into ordeal'
    )} }]->(ordeal)`,
    `(ordeal)-[:PROCESS { reason: ${ggQuote(
      'transformation is paid, never free'
    )} }]->(boon)`,
    `(home | boon | void_boundary)-[:FOLD { strategy: ${ggQuote(
      'reentry-with-lowered-deficit'
    )} }]->(reentry_arc)`,
    '(reentry_arc)-[:PROCESS]->(film)',
    '(film)-[:PROCESS]->(result)',
    '',
  ].join('\n');
}

function buildTestSuite(entries: RankedFilm[]): string {
  const pages = new Map<number, RankedFilm[]>();
  for (const entry of entries) {
    const bucket = pages.get(entry.page) ?? [];
    bucket.push(entry);
    pages.set(entry.page, bucket);
  }

  const orderedPages = [...pages.entries()].sort(([a], [b]) => a - b);
  const pageIds = orderedPages.map(([page]) => `page_${page}`);
  const pageVerdictIds = orderedPages.map(([page]) => `page_${page}_verdict`);

  const lines: string[] = [
    `// imdb-top-250-monomyths.test.gg -- verify the full IMDb Top 250 monomyth corpus (${SNAPSHOT_DATE})`,
    '// Page buckets mirror the source pagination from List Challenges.',
    '',
    `(test_suite: TestSuite { name: ${ggQuote(
      `imdb-top-250-monomyths-${SNAPSHOT_DATE}`
    )} })`,
  ];

  for (const [page, bucket] of orderedPages) {
    bucket.sort((a, b) => a.rank - b.rank);
    lines.push(
      `(page_${page}: Batch { page: ${ggQuote(String(page))}, range: ${ggQuote(
        `${padRank(bucket[0]!.rank)}-${padRank(
          bucket[bucket.length - 1]!.rank
        )}`
      )}, source: ${ggQuote(getPageUrl(page))} })`
    );
    for (const entry of bucket) {
      lines.push(
        `(verify_${padRank(entry.rank)}: Verify { module: ${ggSingleQuote(
          `./${getFileName(entry)}`
        )}, beta1_max: ${ggSingleQuote('8')} })`
      );
    }
  }

  lines.push('');
  lines.push(
    `(test_suite)-[:FORK { reason: ${ggQuote(
      'verify every source page of the static IMDb snapshot'
    )} }]->(${pageIds.join(' | ')})`
  );

  for (const [page, bucket] of orderedPages) {
    const verifyIds = bucket.map((entry) => `verify_${padRank(entry.rank)}`);
    const resultIds = bucket.map((entry) => `result_${padRank(entry.rank)}`);

    lines.push(
      `(page_${page})-[:FORK { reason: ${ggQuote(
        `verify page ${page} witnesses`
      )} }]->(${verifyIds.join(' | ')})`
    );
    for (const entry of bucket) {
      lines.push(
        `(verify_${padRank(entry.rank)})-[:RACE { expect: ${ggQuote(
          'safe'
        )} }]->(result_${padRank(entry.rank)}: Result)`
      );
    }
    lines.push(
      `(${resultIds.join(' | ')})-[:FOLD { strategy: ${ggQuote(
        'all-pass'
      )} }]->(page_${page}_verdict: Verdict)`
    );
  }

  lines.push(
    `(${pageVerdictIds.join(' | ')})-[:FOLD { strategy: ${ggQuote(
      'all-pass'
    )} }]->(verdict: Verdict)`,
    ''
  );

  return lines.join('\n');
}

function cleanGeneratedFiles(): void {
  mkdirSync(OUTPUT_DIR, { recursive: true });
  for (const entry of readdirSync(OUTPUT_DIR)) {
    if (
      /^\d{3}-.*\.gg$/u.test(entry) ||
      entry === 'imdb-top-250-monomyths.test.gg'
    ) {
      rmSync(join(OUTPUT_DIR, entry));
    }
  }
}

async function collectEntries(): Promise<RankedFilm[]> {
  const entries: RankedFilm[] = [];

  for (let page = 1; page <= 7; page += 1) {
    const sourceUrl = getPageUrl(page);
    const html = await downloadText(sourceUrl);
    entries.push(...parsePage(html, page, sourceUrl));
  }

  entries.sort((a, b) => a.rank - b.rank);
  validateEntries(entries);
  return entries;
}

/**
 * Runs the gnosis command-line workflow.
 */
export async function main(): Promise<void> {
  console.log(
    `Generating IMDb Top 250 monomyth corpus from ${BASE_URL} (${SNAPSHOT_DATE})`
  );

  const entries = await collectEntries();
  cleanGeneratedFiles();

  for (const entry of entries) {
    writeFileSync(join(OUTPUT_DIR, getFileName(entry)), buildModule(entry));
  }

  writeFileSync(TEST_FILE, buildTestSuite(entries));

  console.log(`Wrote ${entries.length} GG witnesses to ${OUTPUT_DIR}`);
  console.log(`Wrote test suite: ${TEST_FILE}`);
}

if (process.argv[1] && fileURLToPath(import.meta.url) === process.argv[1]) {
  await main();
}
