// @a0n/gnosis/scripts/validate-mutation-theorem-map.ts
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import { join } from 'node:path';
import { pathToFileURL } from 'node:url';

interface MutationTheoremMap {
  version: number;
  byMutatorName: Record<string, string[]>;
  byPathPrefix: Record<string, string[]>;
}

function loadMutationMap(path: string): MutationTheoremMap {
  const raw = readFileSync(path, 'utf8');
  const parsed: unknown = JSON.parse(raw);
  assert.ok(parsed && typeof parsed === 'object', 'mutation map must be an object');
  const obj = parsed as Record<string, unknown>;
  assert.ok(
    typeof obj.byMutatorName === 'object' &&
      obj.byMutatorName !== null &&
      !Array.isArray(obj.byMutatorName),
    'mutation map must have byMutatorName object'
  );
  assert.ok(
    typeof obj.byPathPrefix === 'object' &&
      obj.byPathPrefix !== null &&
      !Array.isArray(obj.byPathPrefix),
    'mutation map must have byPathPrefix object'
  );
  return parsed as MutationTheoremMap;
}

function collectTheoremIds(map: MutationTheoremMap): Set<string> {
  const ids = new Set<string>();
  for (const arr of Object.values(map.byMutatorName)) {
    for (const id of arr) {
      ids.add(id);
    }
  }
  for (const arr of Object.values(map.byPathPrefix)) {
    for (const id of arr) {
      ids.add(id);
    }
  }
  return ids;
}

/** Ledger may cite an id as `` `THM-…` `` or as plain `THM-…`. */
function theoremIdReferencedInLedger(ledger: string, id: string): boolean {
  if (ledger.includes(`\`${id}\``)) {
    return true;
  }
  return ledger.includes(id);
}

/**
 * Runs the gnosis command-line workflow.
 */
export function main(): void {
  const gnosisRoot = process.cwd();
  const mapPath = join(
    gnosisRoot,
    'src/testing/data/mutation-theorem-map.json'
  );
  const ledgerPath = join(gnosisRoot, 'THEOREM_LEDGER.md');

  const map = loadMutationMap(mapPath);
  const ledger = readFileSync(ledgerPath, 'utf8');
  const ids = collectTheoremIds(map);

  const missing = [...ids]
    .filter((id) => !theoremIdReferencedInLedger(ledger, id))
    .sort();

  if (missing.length > 0) {
    process.stderr.write(
      'Theorem IDs from mutation-theorem-map.json missing from THEOREM_LEDGER.md:\n'
    );
    for (const id of missing) {
      process.stderr.write(`  - ${id}\n`);
    }
  }

  assert.deepStrictEqual(
    missing,
    [],
    'Every mutation-map theorem id must appear in THEOREM_LEDGER.md'
  );
}

const entry = process.argv[1];
if (entry !== undefined && import.meta.url === pathToFileURL(entry).href) {
  main();
}
