// @a0n/.ignored_gnosis/src/ditto/access-trace.ts
/**
 * AccessTrace -- records every interaction with a Chameleon proxy.
 *
 * Functions that receive a Chameleon don't need to know its shape in advance.
 * As the function runs, every `.prop`, `[i]`, `()`, `valueOf`, iteration, and
 * coercion against the chameleon is recorded. The trace becomes a shape map:
 * "to satisfy this function, provide these paths with these types."
 *
 * Traces feed the chaos probe: a population of ChameleonSpecs is evolved to
 * cover all distinct access patterns seen across a FORK's branches.
 */

export type AccessKind =
  | 'get'          // obj.prop or obj['prop']
  | 'has'          // 'prop' in obj
  | 'call'         // obj(...)
  | 'construct'    // new obj(...)
  | 'iterate'      // for-of / ... spread
  | 'valueOf'      // numeric coercion
  | 'toString'     // string coercion
  | 'toPrimitive'  // Symbol.toPrimitive
  | 'keys';        // Object.keys / for-in

export interface AccessEntry {
  /** Access path from the root chameleon, e.g. ".state.godVM.c1" */
  readonly path: string;
  readonly kind: AccessKind;
  /** For 'get': the property name. For 'call': arg count. For 'valueOf': the hint. */
  readonly detail: string | number | null;
  /** Monotonic counter so traces can be diffed by order */
  readonly sequence: number;
}

export class AccessTrace {
  private entries: AccessEntry[] = [];
  private seq = 0;

  record(path: string, kind: AccessKind, detail: string | number | null = null): void {
    this.entries.push({ path, kind, detail, sequence: this.seq++ });
  }

  all(): readonly AccessEntry[] {
    return this.entries;
  }

  /** Unique property paths the function reached. */
  paths(): string[] {
    const seen = new Set<string>();
    for (const e of this.entries) {
      if (e.kind === 'get' || e.kind === 'has') seen.add(e.path);
    }
    return [...seen];
  }

  /** Max depth of any accessed path. */
  maxDepth(): number {
    let max = 0;
    for (const e of this.entries) {
      const depth = e.path.split('.').length - 1;
      if (depth > max) max = depth;
    }
    return max;
  }

  /** Clear and reset. */
  reset(): void {
    this.entries = [];
    this.seq = 0;
  }

  toJSON(): object {
    return {
      entries: this.entries,
      pathCount: this.paths().length,
      maxDepth: this.maxDepth(),
      totalAccesses: this.entries.length,
    };
  }
}
