// @a0n/.ignored_gnosis/src/ditto/chameleon.ts
/**
 * Chameleon -- a universal proxy mock that takes whatever shape the code
 * requests at runtime.
 *
 * A Chameleon is a Proxy. Reading any property returns another Chameleon.
 * Arithmetic coerces it to a configurable number. String coercion returns
 * a configurable string. Iteration returns a configurable array. Calling
 * it returns another Chameleon. The goal: let unfamiliar code execute
 * without pre-knowing its argument shape, while recording every access so
 * we can replay the run with a concrete payload that produces identical
 * (or intentionally divergent) behavior.
 *
 * Pass a `ChameleonSpec` to pin specific paths to specific values -- those
 * paths return the pinned value instead of a chameleon. That's how the
 * chaos probe steers execution down a particular FORK branch.
 */

import { AccessTrace } from './access-trace';

/**
 * A ChameleonValue is anything that can be substituted at a specific access
 * path. Primitives and arrays are returned directly; plain objects become
 * chameleons themselves (nested paths continue to work).
 */
export type ChameleonValue =
  | string
  | number
  | boolean
  | null
  | undefined
  | readonly unknown[]
  | Record<string, unknown>;

export interface ChameleonDefaults {
  readonly number: number;
  readonly string: string;
  readonly boolean: boolean;
  readonly iterableLength: number;
}

export interface ChameleonSpec {
  /** Pinned paths: ".state.godVM.c1" → 1, ".symbol" → "BTC/USD", etc. */
  readonly paths?: Readonly<Record<string, ChameleonValue>>;
  /** Defaults used when a pin doesn't match. */
  readonly defaults?: Partial<ChameleonDefaults>;
  /** Override: if present, calling the chameleon returns this (string coerced). */
  readonly callReturn?: ChameleonValue;
}

const DEFAULT_DEFAULTS: ChameleonDefaults = {
  number: 0,
  string: '',
  boolean: false,
  iterableLength: 0,
};

function joinPath(base: string, key: string | symbol): string {
  if (typeof key === 'symbol') return `${base}[${String(key)}]`;
  if (/^\d+$/.test(key)) return `${base}[${key}]`;
  return base === '' ? `.${key}` : `${base}.${key}`;
}

/**
 * Build a chameleon proxy plus its access trace. The proxy is safe to pass
 * as any argument to any function; the trace records everything.
 */
export function makeChameleon(
  spec: ChameleonSpec = {}
): { proxy: unknown; trace: AccessTrace; spec: ChameleonSpec } {
  const trace = new AccessTrace();
  const defaults: ChameleonDefaults = { ...DEFAULT_DEFAULTS, ...spec.defaults };
  const pinned = spec.paths ?? {};

  // Callable chameleon must wrap a function so the `apply` trap fires.
  // Use a function as the target so typeof returns 'function' when checked.
  const proxy = buildChameleonAt('', trace, defaults, pinned, spec.callReturn);
  return { proxy, trace, spec };
}

function buildChameleonAt(
  path: string,
  trace: AccessTrace,
  defaults: ChameleonDefaults,
  pinned: Readonly<Record<string, ChameleonValue>>,
  callReturn: ChameleonValue | undefined
): unknown {
  const target = function chameleon() {
    // body is never invoked directly; apply trap handles it
  } as unknown as object;

  const handler: ProxyHandler<object> = {
    get(_t, key, receiver) {
      // Symbol.toPrimitive / valueOf / toString drive coercion
      if (key === Symbol.toPrimitive) {
        return (hint: string) => {
          trace.record(path, 'toPrimitive', hint);
          if (hint === 'number') return defaults.number;
          if (hint === 'string') return defaults.string;
          // hint === 'default'
          return defaults.number;
        };
      }
      if (key === 'valueOf') {
        return () => {
          trace.record(path, 'valueOf', null);
          return defaults.number;
        };
      }
      if (key === 'toString') {
        return () => {
          trace.record(path, 'toString', null);
          return defaults.string || `[Chameleon ${path || 'root'}]`;
        };
      }
      if (key === Symbol.iterator) {
        return function* () {
          trace.record(path, 'iterate', null);
          for (let i = 0; i < defaults.iterableLength; i++) {
            yield buildChameleonAt(
              joinPath(path, String(i)),
              trace,
              defaults,
              pinned,
              callReturn
            );
          }
        };
      }
      // Common hot paths we should short-circuit so Node doesn't explode
      if (key === 'then') {
        // Prevent accidental await-as-promise traps
        return undefined;
      }
      if (key === Symbol.asyncIterator) {
        return undefined;
      }
      if (typeof key === 'symbol') {
        // Unknown symbols -> undefined rather than recursive chameleon
        // (prevents issues with e.g. util.inspect.custom)
        return undefined;
      }

      const nextPath = joinPath(path, key);
      trace.record(nextPath, 'get', key);

      // Pinned path override
      if (nextPath in pinned) {
        return materializePinned(pinned[nextPath]);
      }

      // Numeric length-ish properties should return number defaults
      if (key === 'length' || key === 'size' || key === 'byteLength') {
        return defaults.iterableLength;
      }

      // Recurse
      return buildChameleonAt(nextPath, trace, defaults, pinned, callReturn);
    },

    has(_t, key) {
      if (typeof key === 'symbol') return false;
      const nextPath = joinPath(path, key);
      trace.record(nextPath, 'has', key);
      return nextPath in pinned || true;
    },

    apply(_t, _thisArg, args) {
      trace.record(path, 'call', args.length);
      if (callReturn !== undefined) {
        return materializePinned(callReturn);
      }
      return buildChameleonAt(
        path + '()',
        trace,
        defaults,
        pinned,
        callReturn
      );
    },

    construct(_t, args) {
      trace.record(path, 'construct', args.length);
      return buildChameleonAt(
        path + '(new)',
        trace,
        defaults,
        pinned,
        callReturn
      ) as object;
    },

    ownKeys() {
      // Return the pinned keys that start with this path
      const prefix = path === '' ? '.' : path + '.';
      const keys = Object.keys(pinned)
        .filter((p) => p.startsWith(prefix))
        .map((p) => p.slice(prefix.length).split('.')[0]);
      trace.record(path, 'keys', keys.length);
      return [...new Set(keys)];
    },

    getOwnPropertyDescriptor(_t, key) {
      if (typeof key === 'symbol') return undefined;
      return {
        enumerable: true,
        configurable: true,
        writable: true,
        value: buildChameleonAt(
          joinPath(path, key),
          trace,
          defaults,
          pinned,
          callReturn
        ),
      };
    },
  };

  return new Proxy(target, handler);
}

/**
 * Turn a pinned value into what gets returned from the proxy. Objects and
 * arrays are returned as-is; primitives as-is. (Nested chameleons are not
 * automatically synthesized from plain objects -- if the function needs
 * to read deep into a pinned object, the caller should provide a nested
 * path in the spec rather than nesting an object.)
 */
function materializePinned(value: ChameleonValue): unknown {
  return value;
}

/**
 * Convenience: run a function with a chameleon arg and return the trace
 * alongside the result. Errors thrown by the function are captured but
 * not re-raised -- speculative execution tolerates crashes, since the
 * goal is to observe which paths were touched before the crash.
 */
export function runWithChameleon<T>(
  fn: (arg: unknown) => T,
  spec: ChameleonSpec = {}
): { result: T | undefined; trace: AccessTrace; error: Error | null } {
  const { proxy, trace } = makeChameleon(spec);
  try {
    const result = fn(proxy);
    return { result, trace, error: null };
  } catch (e) {
    return {
      result: undefined,
      trace,
      error: e instanceof Error ? e : new Error(String(e)),
    };
  }
}
