// @a0n/.ignored_gnosis/src/ditto/chaos-probe.ts
/**
 * Chaos Probe -- evolutionary search for ChameleonSpecs that collectively
 * cover all FORK branches of a target function.
 *
 * Three lanes run concurrently per agent (quantum fork/race/fold):
 *   Lane A: symbolic prediction
 *   Lane B: runtime observation with chameleon
 *   Lane C: mutation falsification
 *
 * Per generation: evaluate all agents across lanes (in parallel via
 * Promise.all -- the work is CPU-bound but small), select winners via
 * the God Formula, crossover and mutate to refill the population,
 * repeat until every branch is covered or fitness plateaus.
 *
 * Output: a minimal set of ChameleonSpecs that together hit every branch,
 * each tagged with which branches it covers. These become test inputs.
 */

import type { BranchSpec, FitnessInput } from './fitness';
import {
  foldFitness,
  runLaneA,
  runLaneB,
  runLaneC,
  type CompositeFitness,
} from './fitness';
import type { ChameleonSpec, ChameleonValue } from './chameleon';
import {
  crossover,
  godWeight,
  makeRng,
  mutate,
  seedPopulation,
  tournamentSelect,
  type ScoredGenome,
  type XorShift,
} from './gene';

export interface ProbeOptions {
  readonly populationSize?: number;   // default 16
  readonly generations?: number;       // default 6
  readonly elitism?: number;           // default 4
  readonly seed?: number;              // default Date.now()
  readonly plateauGenerations?: number; // early-stop if fitness doesn't improve (default 3)
}

export interface ProbeResult {
  readonly covered: Set<number>;
  readonly finalPopulation: readonly ScoredGenome[];
  readonly generationsRun: number;
  readonly winners: readonly WinnerSpec[];
  readonly coverageTimeline: readonly number[];
}

export interface WinnerSpec {
  readonly spec: ChameleonSpec;
  readonly branchesCovered: readonly number[];
  readonly fitness: CompositeFitness;
}

/**
 * Run the chaos probe on a single function.
 *
 * The caller supplies:
 *   - `branches`:    branch descriptors (index, condition, testsParam)
 *   - `input`:       fitness evaluation hooks (run + argsForSpec)
 *   - `knownValues`: a starting pool of candidate values per path
 *                    (derived from condition RHS literals, types, defaults)
 *   - `mutants`:     optional lane C mutants; if empty, lane C is skipped
 */
export function runChaosProbe(config: {
  branches: readonly BranchSpec[];
  input: FitnessInput;
  knownValues: ReadonlyMap<string, readonly ChameleonValue[]>;
  mutants?: ReadonlyArray<{
    branchIndex: number;
    run: FitnessInput['run'];
  }>;
  options?: ProbeOptions;
}): ProbeResult {
  const opts: Required<ProbeOptions> = {
    populationSize: config.options?.populationSize ?? 16,
    generations: config.options?.generations ?? 6,
    elitism: config.options?.elitism ?? 4,
    seed: config.options?.seed ?? Date.now(),
    plateauGenerations: config.options?.plateauGenerations ?? 3,
  };

  const rng = makeRng(opts.seed);
  const knownPaths = [...config.knownValues.keys()];
  const mutants = config.mutants ?? [];

  // Seed the initial population
  let population: ScoredGenome[] = seedPopulation({
    knownPaths,
    knownValues: config.knownValues,
    populationSize: opts.populationSize,
    rng,
  }).map((g) => ({ genome: g, fitness: 0, rejections: 0 }));

  const coverageTimeline: number[] = [];
  let bestCoverage = new Set<number>();
  let platCounter = 0;
  let generationsRun = 0;

  for (let gen = 0; gen < opts.generations; gen++) {
    generationsRun++;

    // Evaluate fitness for every genome across three lanes
    const scored = evaluateGeneration({
      population,
      config,
      mutants,
    });

    // Track aggregate coverage
    const genCoverage = new Set<number>();
    for (const s of scored) {
      for (const b of s.coverage) genCoverage.add(b);
    }
    coverageTimeline.push(genCoverage.size);

    // Early stop: all branches covered
    if (genCoverage.size >= config.branches.length) {
      population = scored.map((s) => ({
        genome: s.genome,
        fitness: s.fitness.total,
        rejections: 0,
      }));
      bestCoverage = genCoverage;
      break;
    }

    // Plateau detection
    if (genCoverage.size > bestCoverage.size) {
      bestCoverage = genCoverage;
      platCounter = 0;
    } else {
      platCounter++;
      if (platCounter >= opts.plateauGenerations) {
        population = scored.map((s) => ({
          genome: s.genome,
          fitness: s.fitness.total,
          rejections: 0,
        }));
        break;
      }
    }

    // Track rejections (generations without per-agent fitness improvement)
    const prevFitness = new Map(
      population.map((p) => [hashSpec(p.genome), p.fitness])
    );
    population = scored.map((s) => {
      const prev = prevFitness.get(hashSpec(s.genome)) ?? 0;
      const rejected = s.fitness.total <= prev ? 1 : 0;
      return {
        genome: s.genome,
        fitness: s.fitness.total,
        rejections:
          (population.find((p) => hashSpec(p.genome) === hashSpec(s.genome))
            ?.rejections ?? 0) + rejected,
      };
    });

    // Select winners via God Formula
    const elites = tournamentSelect(population, opts.elitism);

    // Refill population: elitism + crossover + mutation
    const next: ScoredGenome[] = [...elites];
    while (next.length < opts.populationSize) {
      const parentA = rng.pick(elites).genome;
      const parentB = rng.pick(population).genome;
      let child = crossover(parentA, parentB, rng);
      if (rng.bool(0.4)) {
        child = mutate(child, {
          knownPaths,
          knownValues: config.knownValues,
          rng,
        });
      }
      next.push({ genome: child, fitness: 0, rejections: 0 });
    }
    population = next;
  }

  // Build the winner set: minimal cover via greedy selection
  const winners = greedyMinimalCover(
    evaluateGeneration({ population, config, mutants }),
    config.branches.length
  );

  return {
    covered: bestCoverage,
    finalPopulation: population,
    generationsRun,
    winners,
    coverageTimeline,
  };
}

// ---------------------------------------------------------------------------
// Per-generation evaluation (three lanes in parallel per agent)
// ---------------------------------------------------------------------------

interface ScoredEval {
  readonly genome: ChameleonSpec;
  readonly fitness: CompositeFitness;
  readonly coverage: Set<number>;
}

function evaluateGeneration(args: {
  population: readonly ScoredGenome[];
  config: {
    branches: readonly BranchSpec[];
    input: FitnessInput;
    knownValues: ReadonlyMap<string, readonly ChameleonValue[]>;
  };
  mutants: ReadonlyArray<{
    branchIndex: number;
    run: FitnessInput['run'];
  }>;
}): ScoredEval[] {
  return args.population.map(({ genome }) => {
    // Lane A: symbolic
    const a = runLaneA(args.config.branches, genome);

    // Lane B: runtime
    const b = runLaneB(args.config.input, genome);

    // Lane C: falsification (only on branches hit by A or B)
    const cCandidates = args.mutants.filter(
      (m) => a.decided.has(m.branchIndex) || b.hits.has(m.branchIndex)
    );
    const c = runLaneC({
      baseline: b,
      branches: args.config.branches,
      mutants: cCandidates,
      argsForSpec: args.config.input.argsForSpec,
      spec: genome,
    });

    const fitness = foldFitness({
      branchCount: args.config.branches.length,
      a,
      b,
      c,
    });
    return { genome, fitness, coverage: fitness.coverage };
  });
}

// ---------------------------------------------------------------------------
// Greedy minimal cover: pick the smallest set of specs that covers all branches
// ---------------------------------------------------------------------------

function greedyMinimalCover(
  evaluated: readonly ScoredEval[],
  branchCount: number
): WinnerSpec[] {
  const winners: WinnerSpec[] = [];
  const covered = new Set<number>();
  const candidates = [...evaluated].sort(
    (a, b) => b.coverage.size - a.coverage.size
  );

  for (const cand of candidates) {
    if (covered.size >= branchCount) break;
    const newHits = [...cand.coverage].filter((x) => !covered.has(x));
    if (newHits.length === 0) continue;
    for (const x of newHits) covered.add(x);
    winners.push({
      spec: cand.genome,
      branchesCovered: newHits,
      fitness: cand.fitness,
    });
  }

  return winners;
}

function hashSpec(spec: ChameleonSpec): string {
  const paths = spec.paths ?? {};
  return Object.keys(paths)
    .sort()
    .map((k) => `${k}=${JSON.stringify(paths[k])}`)
    .join('|');
}

// ---------------------------------------------------------------------------
// Re-exports for convenience
// ---------------------------------------------------------------------------

export { godWeight };
export type { XorShift };
