// @a0n/.ignored_gnosis/src/__tests__/betti-wasm-codegen.test.ts
/**
 * Betti → WASM worker-side artifact map codegen tests: deterministic output
 * (sorted by artifactId, byte-identical across runs and input orders),
 * correct import/map/manifest rendering, module copies, and the fail-closed
 * paths (missing sidecar, tampered bytes, conflicting duplicates).
 */
import { createHash } from 'node:crypto';
import { mkdtemp, readFile, rm, writeFile } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { describe, expect, test } from '@a0n/gnosis/test';
import type { BettiWasmManifest } from '../betti-wasm-manifest';
import {
  BETTI_ARTIFACTS_MODULE_HEADER,
  generateBettiArtifactsModule,
  renderBettiArtifactsModule,
} from '../betti-wasm-codegen';

// Hand-assembled scalar-v1 module (doubles its argument) — same fixture as
// betti-wasm-store.test.ts.
const SCALAR_DOUBLE_WASM = Uint8Array.from([
  0x00, 0x61, 0x73, 0x6d, 0x01, 0x00, 0x00, 0x00,
  0x01, 0x06, 0x01, 0x60, 0x01, 0x7c, 0x01, 0x7c,
  0x03, 0x02, 0x01, 0x00,
  0x07, 0x08, 0x01, 0x04, 0x6d, 0x61, 0x69, 0x6e, 0x00, 0x00,
  0x0a, 0x10, 0x01, 0x0e, 0x00, 0x20, 0x00, 0x44,
  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x40,
  0xa2, 0x0b,
]);

function makeManifest(artifactId: string, bytes: Uint8Array): BettiWasmManifest {
  return {
    artifact_version: 'betti-wasm/1',
    artifactId,
    topologyHash: 'aaaa5555aaaa5555',
    sourceHash: '5555aaaa5555aaaa',
    language: 'typescript',
    filePath: `gnode/benchmarks/${artifactId}.ts`,
    functionName: 'main',
    certified_class: 'numeric-loop',
    tier: 'certified-direct',
    export_name: 'main',
    abi: 'scalar-v1',
    imports: [],
    pure: true,
    limits: { memoryPagesMax: 1 },
    byteLength: bytes.length,
    sha256: createHash('sha256').update(bytes).digest('hex'),
    decline_semantics: 'trap',
    wasm_opt: false,
  } as BettiWasmManifest;
}

async function writePair(
  dir: string,
  name: string,
  artifactId: string,
  bytes: Uint8Array = SCALAR_DOUBLE_WASM
): Promise<BettiWasmManifest> {
  const manifest = makeManifest(artifactId, bytes);
  await writeFile(join(dir, `${name}.wasm`), bytes);
  await writeFile(
    join(dir, `${name}.manifest.json`),
    JSON.stringify(manifest, null, 2),
    'utf8'
  );
  return manifest;
}

describe('betti wasm codegen', () => {
  test('renderBettiArtifactsModule is order-independent and sorted', () => {
    const a = makeManifest('aaaaaaaaaaaaaaaa', SCALAR_DOUBLE_WASM);
    const b = makeManifest('bbbbbbbbbbbbbbbb', SCALAR_DOUBLE_WASM);
    const forward = renderBettiArtifactsModule([a, b]);
    const reversed = renderBettiArtifactsModule([b, a]);
    expect(forward).toBe(reversed);

    const lines = forward.split('\n');
    expect(lines[0]).toBe(BETTI_ARTIFACTS_MODULE_HEADER);
    const importLines = lines.filter((line) => line.startsWith('import '));
    expect(importLines).toEqual([
      "import m_aaaaaaaaaaaaaaaa from './wasm/aaaaaaaaaaaaaaaa.wasm';",
      "import m_bbbbbbbbbbbbbbbb from './wasm/bbbbbbbbbbbbbbbb.wasm';",
    ]);
    expect(forward).toContain(
      "  ['aaaaaaaaaaaaaaaa', m_aaaaaaaaaaaaaaaa as unknown as WebAssembly.Module],"
    );
    expect(forward).toContain('export const bettiManifests = [');
    expect(forward).toContain('] as const;');
  });

  test('generates a deterministic module with copied wasm files', async () => {
    const artifactsDir = await mkdtemp(join(tmpdir(), 'betti-codegen-src-'));
    const outDirA = await mkdtemp(join(tmpdir(), 'betti-codegen-out-a-'));
    const outDirB = await mkdtemp(join(tmpdir(), 'betti-codegen-out-b-'));
    try {
      // Write out of artifactId order — output must sort by artifactId.
      await writePair(artifactsDir, 'zeta', 'bbbbbbbbbbbbbbbb');
      await writePair(artifactsDir, 'alpha', 'aaaaaaaaaaaaaaaa');

      const first = await generateBettiArtifactsModule({
        artifactsDir,
        outFile: join(outDirA, 'betti-artifacts.ts'),
      });
      expect(first.artifacts.map((artifact) => artifact.artifactId)).toEqual([
        'aaaaaaaaaaaaaaaa',
        'bbbbbbbbbbbbbbbb',
      ]);

      const copied = new Uint8Array(
        await readFile(join(outDirA, 'wasm', 'aaaaaaaaaaaaaaaa.wasm'))
      );
      expect(copied.length).toBe(SCALAR_DOUBLE_WASM.length);
      expect([...copied]).toEqual([...SCALAR_DOUBLE_WASM]);

      const moduleText = await readFile(first.outFile, 'utf8');
      const manifests = JSON.parse(
        `[${moduleText
          .slice(
            moduleText.indexOf('export const bettiManifests = [') +
              'export const bettiManifests = ['.length,
            moduleText.lastIndexOf('] as const;')
          )
          .trim()
          .replace(/,\s*$/, '')}]`
      ) as { artifactId: string }[];
      expect(manifests.map((manifest) => manifest.artifactId)).toEqual([
        'aaaaaaaaaaaaaaaa',
        'bbbbbbbbbbbbbbbb',
      ]);

      const declaration = await readFile(first.declarationFile, 'utf8');
      expect(declaration).toContain("declare module '*.wasm'");

      // Second run into a fresh directory → byte-identical module text.
      const second = await generateBettiArtifactsModule({
        artifactsDir,
        outFile: join(outDirB, 'betti-artifacts.ts'),
      });
      expect(await readFile(second.outFile, 'utf8')).toBe(moduleText);
    } finally {
      await rm(artifactsDir, { recursive: true, force: true });
      await rm(outDirA, { recursive: true, force: true });
      await rm(outDirB, { recursive: true, force: true });
    }
  });

  test('fails closed on a missing manifest sidecar', async () => {
    const artifactsDir = await mkdtemp(join(tmpdir(), 'betti-codegen-src-'));
    const outDir = await mkdtemp(join(tmpdir(), 'betti-codegen-out-'));
    try {
      await writeFile(join(artifactsDir, 'orphan.wasm'), SCALAR_DOUBLE_WASM);
      let failure: unknown = null;
      try {
        await generateBettiArtifactsModule({
          artifactsDir,
          outFile: join(outDir, 'betti-artifacts.ts'),
        });
      } catch (error) {
        failure = error;
      }
      expect(String(failure)).toContain('no manifest sidecar');
    } finally {
      await rm(artifactsDir, { recursive: true, force: true });
      await rm(outDir, { recursive: true, force: true });
    }
  });

  test('fails closed on tampered module bytes', async () => {
    const artifactsDir = await mkdtemp(join(tmpdir(), 'betti-codegen-src-'));
    const outDir = await mkdtemp(join(tmpdir(), 'betti-codegen-out-'));
    try {
      const manifest = makeManifest('aaaaaaaaaaaaaaaa', SCALAR_DOUBLE_WASM);
      const tampered = Uint8Array.from(SCALAR_DOUBLE_WASM);
      tampered[18] ^= 0x01;
      await writeFile(join(artifactsDir, 'bad.wasm'), tampered);
      await writeFile(
        join(artifactsDir, 'bad.manifest.json'),
        JSON.stringify(manifest, null, 2),
        'utf8'
      );

      let failure: unknown = null;
      try {
        await generateBettiArtifactsModule({
          artifactsDir,
          outFile: join(outDir, 'betti-artifacts.ts'),
        });
      } catch (error) {
        failure = error;
      }
      expect(String(failure)).toContain('integrity failure');
    } finally {
      await rm(artifactsDir, { recursive: true, force: true });
      await rm(outDir, { recursive: true, force: true });
    }
  });

  test('identical duplicates collapse; conflicting duplicates fail closed', async () => {
    const artifactsDir = await mkdtemp(join(tmpdir(), 'betti-codegen-src-'));
    const outDir = await mkdtemp(join(tmpdir(), 'betti-codegen-out-'));
    try {
      await writePair(artifactsDir, 'copy-one', 'aaaaaaaaaaaaaaaa');
      await writePair(artifactsDir, 'copy-two', 'aaaaaaaaaaaaaaaa');
      const collapsed = await generateBettiArtifactsModule({
        artifactsDir,
        outFile: join(outDir, 'betti-artifacts.ts'),
      });
      expect(collapsed.artifacts.length).toBe(1);

      // Same artifactId, different bytes (still self-consistent pair) → throw.
      const conflicting = Uint8Array.from(SCALAR_DOUBLE_WASM);
      conflicting[46] = 0x0b; // stays a different byte stream
      conflicting[45] = 0xa0; // f64.sub instead of f64.mul
      await writePair(artifactsDir, 'copy-three', 'aaaaaaaaaaaaaaaa', conflicting);

      let failure: unknown = null;
      try {
        await generateBettiArtifactsModule({
          artifactsDir,
          outFile: join(outDir, 'betti-artifacts.ts'),
        });
      } catch (error) {
        failure = error;
      }
      expect(String(failure)).toContain('conflicting duplicates');
    } finally {
      await rm(artifactsDir, { recursive: true, force: true });
      await rm(outDir, { recursive: true, force: true });
    }
  });
});
