// @a0n/.ignored_gnosis/src/__tests__/mesh-policy-braid-fold.test.ts
import { describe, expect, it } from '@a0n/gnosis/test';
import {
  Verdict,
  isAccept,
  isDefer,
  isReject,
} from '@a0n/bitwise';
import {
  auditBraidFoldCandidate,
  auditDecompositionTripleCandidate,
  actionForPortCoordinate,
  certifyBraidFastPathEquivalence,
  createBraidPacketCertificate,
  createBraidFoldCertificate,
  createCalibratedFluidicPressureBridge,
  createClosedLoopRuntimeMetadataCertificate,
  createClosedArchitectureCertificate,
  createDecompositionTripleCertificate,
  createExactDecompositionTriple,
  createExtendedAeonTransitionCertificate,
  createFluidicRoutingAdmissibilityCertificate,
  createGnosticMetrologyProfile,
  createGnosticLensGeometryCertificate,
  createGnosticRuntimeAllocationPlan,
  createIndexedInterferePortAdmissionCertificate,
  createIndexedInterferePortClosureFamilyAuditCertificate,
  createIndexedInterferePortClosureFamilyCertificate,
  createIndexedInterferePortClosureFamilyRejectionCertificate,
  createGnosisUringInterferePortCertificate,
  createInterferenceStrideCertificate,
  createNaturalPortSystemCertificate,
  createPressureCertifiedExecutionPathCertificate,
  createPressureGradientCertificate,
  createPressureWeightedClosedLoopBenchmarkReportCertificate,
  createPressureWeightedInterferenceStrideCertificate,
  createStatisticalMechanicsAccounting,
  createWankelRuntimeCertificate,
  closedLoopRuntimeMetadataChecksum,
  createVentToForkBoundary,
  decompositionFeedsKenomaSpan,
  decompositionTripleCoordinates,
  expectedLayerForSaturation,
  focusForSaturation,
  infiniteForkRaceFoldVentAction,
  infiniteForkRaceFoldVentCoordinate,
  informationMass,
  leakage,
  leakageMass,
  lensShapeForPortCoordinate,
  keystoneGap,
  nextPortAction,
  observeSaturation,
  EXACT_PRESSURE_CERTIFIED_EXECUTION_PATH_IDENTITY,
  PRESSURE_CERTIFIED_EXECUTION_PATH_STABLE_TIE_ANCHOR_THEOREM_IDS,
  pressureWeightedBenchmarkInterferenceStrideTraceFor,
  pressureWeightedBenchmarkTopologyCoordinatesFor,
  pressureWeightedBenchmarkTotalFor,
  pressureWeightedBenchmarkWeightedStrideTraceFor,
  portCoordinateForAction,
  portStreamIsPeriodic,
  returnsToBase,
  selectCertifiedBraidFastPath,
  selectWankelRuntimeFastPath,
  certifyWankelRuntimeFastPathAggregateCostDominance,
  certifyWankelRuntimeFastPathCostDominance,
  certifyWankelRuntimeFastPathEquivalence,
  certifyWankelRuntimeFastPathRobustEstimatorDominance,
  traceTotal,
  type BraidPacketCertificate,
  type BraidFoldCandidate,
  type BraidFoldCertificate,
  type GnosticLensGeometryCertificate,
  type PressureWeightedInterferenceStrideCertificate,
  type PressureCertifiedExecutionPathCertificate,
  type ClosedLoopRuntimeMetadataCertificate,
  type StatisticalMechanicsAccountingCertificate,
  WANKEL_AEON_CELLS,
  WANKEL_FIVE_FORCE_LIFECYCLE_TRACE,
  WANKEL_RUNTIME_CERTIFICATE_THEOREM_IDS,
  WANKEL_RUNTIME_DEADLINE_PULSE_HOPS,
  visibleMass,
} from '../mesh-policy';

const exactCandidate: BraidFoldCandidate = {
  cycle: 3,
  gateCoordinates: [10, 12, 17, 22],
  saturationTrace: [10, 30, 90],
  closedArchitecture: true,
  endpointsLinked: true,
  baseReady: true,
};

describe('mesh-policy braid fold mirror', () => {
  it('mirrors saturation layer and focus selection', () => {
    expect(expectedLayerForSaturation(10)).toBe('kenoma');
    expect(expectedLayerForSaturation(30)).toBe('primeKeystone');
    expect(expectedLayerForSaturation(90)).toBe('doubleKeystone');
    expect(expectedLayerForSaturation(12)).toBe(null);

    expect(focusForSaturation(30)).toBe('witnessInversion');
    expect(focusForSaturation(90)).toBe('hostHandshake');
    expect(observeSaturation(12).admitted).toBe(false);
  });

  it('creates the closed architecture certificate with base return', () => {
    const certificate = createClosedArchitectureCertificate();

    expect(certificate.layerCoordinates).toEqual([10, 12, 17, 22]);
    expect(certificate.saturationTrace).toEqual([10, 30, 90]);
    expect(certificate.keystoneGap).toBe(5);
    expect(returnsToBase(certificate)).toBe(true);
  });

  it('certifies the natural 10/12/17/22 port system for infinite fork-race-fold-vent', () => {
    const certificate = createNaturalPortSystemCertificate();

    expect(certificate.actionTrace).toEqual(['fork', 'race', 'fold', 'vent']);
    expect(certificate.portCoordinates).toEqual([10, 12, 17, 22]);
    expect(certificate.firstInfiniteCycleCoordinates).toEqual([10, 12, 17, 22]);
    expect(certificate.cycleLength).toBe(4);
    expect(certificate.ventHandsOffToFork).toBe(true);
    expect(certificate.ventToForkBoundary).toEqual({
      source: 'GnosisNaturalPortCycleBoundary',
      outgoingStage: 3,
      incomingStage: 4,
      outgoingAction: 'vent',
      incomingAction: 'fork',
      outgoingCoordinate: 22,
      incomingCoordinate: 10,
      contiguousStages: true,
      theoremIds: certificate.theoremIds,
    });
    expect(certificate.coordinatesMatchBraidGates).toBe(true);
    expect(returnsToBase(certificate)).toBe(true);

    const boundary = createVentToForkBoundary();
    expect(boundary).toEqual(certificate.ventToForkBoundary);
    expect(portCoordinateForAction('fork')).toBe(10);
    expect(portCoordinateForAction('race')).toBe(12);
    expect(portCoordinateForAction('fold')).toBe(17);
    expect(portCoordinateForAction('vent')).toBe(22);
    expect(portCoordinateForAction('interfere')).toBe(43);
    expect(actionForPortCoordinate(10)).toBe('fork');
    expect(actionForPortCoordinate(12)).toBe('race');
    expect(actionForPortCoordinate(17)).toBe('fold');
    expect(actionForPortCoordinate(22)).toBe('vent');
    expect(actionForPortCoordinate(43)).toBe('interfere');
    expect(actionForPortCoordinate(11)).toBe(null);
    expect(nextPortAction('vent')).toBe('fork');
    expect(nextPortAction('interfere')).toBe('interfere');

    const firstTwoCycles = Array.from({ length: 8 }, (_, stage) =>
      infiniteForkRaceFoldVentAction(stage)
    );
    const firstTwoCoordinateCycles = Array.from({ length: 8 }, (_, stage) =>
      infiniteForkRaceFoldVentCoordinate(stage)
    );
    expect(firstTwoCycles).toEqual([
      'fork',
      'race',
      'fold',
      'vent',
      'fork',
      'race',
      'fold',
      'vent',
    ]);
    expect(firstTwoCoordinateCycles).toEqual([10, 12, 17, 22, 10, 12, 17, 22]);
    expect(infiniteForkRaceFoldVentCoordinate(17)).toBe(
      infiniteForkRaceFoldVentCoordinate(21)
    );
    expect(portStreamIsPeriodic(17)).toBe(true);
    expect(certificate.theoremIds).toContain(
      'GamutBraid.exact_natural_port_coordinates'
    );
    expect(certificate.theoremIds).toContain(
      'GamutBraid.infinite_port_coordinate_four_cycle'
    );
    expect(certificate.theoremIds).toContain(
      'GamutBraid.infinite_port_stream_periodic'
    );
    expect(certificate.theoremIds).toContain(
      'GamutBraid.exact_vent_to_fork_boundary'
    );
  });

  it('accepts the exact braid fold candidate', () => {
    const audit = auditBraidFoldCandidate(exactCandidate);
    const certificate = createBraidFoldCertificate(exactCandidate);

    expect(audit.accepted).toBe(true);
    expect(audit.anchorCoordinates).toEqual([10, 17, 22]);
    expect(certificate?.primeDoubleCrossingCount).toBe(5);
    expect(certificate?.forkedAnchorCoordinates).toEqual([10, 17, 22]);
    expect(certificate ? returnsToBase(certificate) : false).toBe(true);

    const fastPath = selectCertifiedBraidFastPath(certificate);
    expect(fastPath.selected).toBe(true);
    expect(fastPath.path?.collapsedBridge).toEqual({
      from: 17,
      to: 22,
      stride: 5,
    });
  });

  it('accepts the exact 7,17,22 decomposition triple on a braid certificate', () => {
    const braidCertificate = createBraidFoldCertificate(exactCandidate);
    const triple = createExactDecompositionTriple();
    const audit = auditDecompositionTripleCandidate(
      {
        cycle: 3,
        coordinates: [7, 17, 22],
      },
      braidCertificate
    );
    const certificate = createDecompositionTripleCertificate(
      {
        cycle: 3,
        coordinates: [7, 17, 22],
      },
      braidCertificate
    );

    expect(decompositionTripleCoordinates(triple)).toEqual([7, 17, 22]);
    expect(decompositionFeedsKenomaSpan(triple)).toBe(true);
    expect(audit.accepted).toBe(true);
    expect(audit.certificate?.coordinates).toEqual([7, 17, 22]);
    expect(certificate?.primeDoubleCrossingCount).toBe(5);
    expect(certificate ? returnsToBase(certificate) : false).toBe(true);
  });

  it('rejects decomposition triples without exact coordinates or braid closure', () => {
    const braidCertificate = createBraidFoldCertificate(exactCandidate);
    const missingSeven = auditDecompositionTripleCandidate(
      {
        cycle: 3,
        coordinates: [10, 17, 22],
      },
      braidCertificate
    );
    const absentBraid = auditDecompositionTripleCandidate(
      {
        cycle: 3,
        coordinates: [7, 17, 22],
      },
      null
    );
    const mismatchedCycle = auditDecompositionTripleCandidate(
      {
        cycle: 4,
        coordinates: [7, 17, 22],
      },
      braidCertificate
    );

    expect(missingSeven.accepted).toBe(false);
    expect(absentBraid.accepted).toBe(false);
    expect(mismatchedCycle.accepted).toBe(false);
  });

  it('rejects missing anchors, malformed crossing shape, and absent closure', () => {
    const missingTen = auditBraidFoldCandidate({
      ...exactCandidate,
      gateCoordinates: [12, 17, 22],
    });
    const missingTwentyTwo = auditBraidFoldCandidate({
      ...exactCandidate,
      gateCoordinates: [10, 12, 17],
    });
    const wrongCrossing = auditBraidFoldCandidate({
      ...exactCandidate,
      gateCoordinates: [10, 12, 17, 23],
    });
    const absentClosure = auditBraidFoldCandidate({
      ...exactCandidate,
      closedArchitecture: false,
    });

    expect(missingTen.accepted).toBe(false);
    expect(missingTen.anchorCoordinates).toEqual([17, 22]);
    expect(missingTwentyTwo.accepted).toBe(false);
    expect(missingTwentyTwo.anchorCoordinates).toEqual([10, 17]);
    expect(wrongCrossing.accepted).toBe(false);
    expect(absentClosure.accepted).toBe(false);
    expect(
      createBraidFoldCertificate({
        ...exactCandidate,
        closedArchitecture: false,
      })
    ).toBe(null);
  });

  it('keeps the 17-to-22 bridge at crossing count five', () => {
    expect(keystoneGap(17, 22)).toBe(5);
    expect(keystoneGap(22, 17)).toBe(0);
  });

  it('computes statistical mechanics accounting from the exact macrostate', () => {
    expect(visibleMass(10)).toBe(10);
    expect(visibleMass(30)).toBe(12);
    expect(visibleMass(90)).toBe(12);
    expect(leakage(10)).toBe(0);
    expect(leakage(30)).toBe(18);
    expect(leakage(90)).toBe(78);
    expect(informationMass(30)).toBe(360);
    expect(leakageMass(90)).toBe(936);

    const accounting = createStatisticalMechanicsAccounting(
      createClosedArchitectureCertificate()
    );

    expect(accounting?.visibleMassTrace).toEqual([10, 12, 12]);
    expect(accounting?.leakageTrace).toEqual([0, 18, 78]);
    expect(accounting?.informationMassTrace).toEqual([120, 360, 1080]);
    expect(accounting?.leakageMassTrace).toEqual([0, 216, 936]);
    expect(accounting?.totalLeakage).toBe(96);
    expect(accounting?.totalInformationMass).toBe(1560);
    expect(accounting ? returnsToBase(accounting) : false).toBe(true);
    expect(traceTotal(accounting?.leakageTrace ?? [])).toBe(96);
  });

  it('derives gnostic metrology units from the accounting certificate', () => {
    const accounting = createStatisticalMechanicsAccounting(
      createClosedArchitectureCertificate()
    );
    const metrology = createGnosticMetrologyProfile(accounting);

    expect(metrology?.massUnit).toBe('gG');
    expect(metrology?.pressureUnit).toBe('SPa');
    expect(metrology?.localCarryingCapacity).toBe(144);
    expect(metrology?.virtualMass).toBe(1416);
    expect(metrology?.totalGnosticMass).toBe(1560);
    expect(metrology?.totalSievePressure).toBe(96);
    expect(metrology?.checksumMass).toBe(1560);
    expect(metrology?.shootGamutPressureRequirement).toBe(96);
    expect(metrology?.brownBufferMass).toBe(1080);
    expect(metrology?.brownLeakageMass).toBe(936);
    expect(metrology?.layerRows[2]).toEqual({
      band: 'brown',
      saturation: 90,
      visibleMass: 12,
      sievePressure: 78,
      gnosticMass: 1080,
      leakageMass: 936,
    });
    expect(createGnosticMetrologyProfile(null)).toBe(null);
  });

  it('packages the periodic port stream as a 1560-mass braid packet', () => {
    const accounting = createStatisticalMechanicsAccounting(
      createClosedArchitectureCertificate()
    );
    const packet = createBraidPacketCertificate(3, accounting);
    const malformedAccounting =
      accounting === null
        ? null
        : ({
            ...accounting,
            totalInformationMass: 1559,
          } as unknown as StatisticalMechanicsAccountingCertificate);

    expect(packet).toMatchObject({
      source: 'GnosisBraidPacketCertificate',
      cycle: 3,
      actionTrace: ['fork', 'race', 'fold', 'vent'],
      portCoordinates: [10, 12, 17, 22],
      saturationTrace: [10, 30, 90],
      informationMassTrace: [120, 360, 1080],
      leakageTrace: [0, 18, 78],
      totalInformationMass: 1560,
      totalLeakage: 96,
      returnsToBase: true,
    });
    expect(packet?.boundary).toEqual(createVentToForkBoundary());
    expect(packet?.accounting).toEqual(accounting);
    expect(packet ? returnsToBase(packet) : false).toBe(true);
    expect(packet?.theoremIds).toContain(
      'GamutBraid.exact_braid_packet_certified'
    );
    expect(packet?.theoremIds).toContain(
      'StatisticalMechanicsAccounting.exact_total_information_mass'
    );
    expect(createBraidPacketCertificate(3, malformedAccounting)).toBe(null);
    expect(createBraidPacketCertificate(-1, accounting)).toBe(null);
    expect(createBraidPacketCertificate(3, null)).toBe(null);
  });

  it('certifies the 43-prime transition as a finite extended-aeon candidate', () => {
    const packet = createBraidPacketCertificate(3);
    const certificate = createExtendedAeonTransitionCertificate(packet);
    const firstClosureFamily =
      createIndexedInterferePortClosureFamilyCertificate(certificate);
    const secondClosureFamily =
      createIndexedInterferePortClosureFamilyCertificate(certificate, 43, 2);
    const secondClosureRejection =
      createIndexedInterferePortClosureFamilyRejectionCertificate(
        secondClosureFamily
      );
    const thirdClosureFamily =
      createIndexedInterferePortClosureFamilyCertificate(certificate, 43, 3);
    const thirdClosureRejection =
      createIndexedInterferePortClosureFamilyRejectionCertificate(
        thirdClosureFamily
      );
    const closureFamilyAudit =
      createIndexedInterferePortClosureFamilyAuditCertificate(certificate);
    const sparseClosureFamilyAudit =
      createIndexedInterferePortClosureFamilyAuditCertificate(
        certificate,
        [1, 4, 5]
      );
    const indexedAdmission =
      createIndexedInterferePortAdmissionCertificate(certificate);
    const interferePort =
      createGnosisUringInterferePortCertificate(certificate);
    const stride = createInterferenceStrideCertificate(packet);
    const pressureWeightedStride =
      createPressureWeightedInterferenceStrideCertificate(stride);
    const malformedPacket =
      packet === null
        ? null
        : ({
            ...packet,
            totalLeakage: 95,
          } as unknown as BraidPacketCertificate);

    expect(certificate).toMatchObject({
      source: 'GnosisExtendedAeonTransitionCertificate',
      ranges: [
        [1, 12],
        [10, 22],
        [20, 32],
        [30, 43],
        [50, 55],
      ],
      keystoneTrace: [
        [7, 11],
        [17, 22],
        [27, 32],
        [37, 43],
        [55, 55],
      ],
      primeTransition: {
        aeon: 'fourth',
        linearTerminal: 42,
        clinamen: 1,
        terminal: 43,
        foldKeystone: 37,
        ventKeystone: 43,
      },
      closure: {
        coordinate: 55,
        carrier: 11,
        cycles: 5,
      },
      returnsToBase: true,
    });
    expect(certificate?.braidPacket).toEqual(packet);
    expect(certificate ? returnsToBase(certificate) : false).toBe(true);
    expect(certificate?.theoremIds).toContain(
      'GamutBraid.exact_fourth_aeon_prime_terminal_transition'
    );
    expect(certificate?.theoremIds).toContain(
      'GamutBraid.exact_fifty_five_ouroboros_closure'
    );
    expect(firstClosureFamily).toMatchObject({
      source: 'GnosisIndexedInterferePortClosureFamilyCertificate',
      portAction: 'interfere',
      portCoordinate: 43,
      closureCoordinate: 55,
      closureIndex: 1,
      closureStride: 12,
      closureStepsAfterExactAdmission: 0,
      closureDistanceFromExactAdmission: 0,
      closureAxisPressureWeightedTotal: 3738,
      closureAxisPressureGrowthFromInterferencePort: 1152,
      closureAxisPressureGrowthFromExactClosure: 0,
      orderedAfterExactAdmission: false,
      exactAdmissionCompatible: true,
      terminal: true,
      returnsToBase: true,
    });
    expect(firstClosureFamily?.theoremIds).toContain(
      'GamutBraid.indexed_interfere_closure_coordinate_strictly_increases'
    );
    expect(firstClosureFamily?.theoremIds).toContain(
      'GamutBraid.indexed_interfere_closure_distance_from_exact_admission'
    );
    expect(firstClosureFamily?.theoremIds).toContain(
      'GamutBraid.indexed_interfere_closure_axis_pressure_total'
    );
    expect(firstClosureFamily?.theoremIds).toContain(
      'GamutBraid.exact_indexed_interfere_port_first_closure_axis_pressure_total'
    );
    expect(firstClosureFamily?.theoremIds).toContain(
      'GamutBraid.exact_indexed_interfere_port_closure_family_member_certified'
    );
    expect(firstClosureFamily?.theoremIds).toContain(
      'GamutBraid.exact_indexed_interfere_port_closure_family_certified'
    );
    expect(secondClosureFamily).toMatchObject({
      source: 'GnosisIndexedInterferePortClosureFamilyCertificate',
      portAction: 'interfere',
      portCoordinate: 43,
      closureCoordinate: 67,
      closureIndex: 2,
      closureStride: 12,
      closureStepsAfterExactAdmission: 1,
      closureDistanceFromExactAdmission: 12,
      closureAxisPressureWeightedTotal: 4890,
      closureAxisPressureGrowthFromInterferencePort: 2304,
      closureAxisPressureGrowthFromExactClosure: 1152,
      orderedAfterExactAdmission: true,
      exactAdmissionCompatible: false,
      terminal: true,
      returnsToBase: true,
    });
    expect(secondClosureFamily?.theoremIds).toContain(
      'GamutBraid.exact_indexed_interfere_port_closure_family_member_certified'
    );
    expect(secondClosureFamily?.theoremIds).toContain(
      'GamutBraid.exact_indexed_interfere_port_second_closure_family_certified'
    );
    expect(secondClosureFamily?.theoremIds).toContain(
      'GamutBraid.exact_indexed_interfere_port_second_closure_family_distance'
    );
    expect(secondClosureFamily?.theoremIds).toContain(
      'GamutBraid.exact_indexed_interfere_port_second_closure_axis_pressure_total'
    );
    expect(secondClosureRejection).toMatchObject({
      source: 'GnosisIndexedInterferePortClosureFamilyRejectionCertificate',
      rejectedBy: 'GnosisIndexedInterferePortAdmissionCertificate',
      reason: 'wrong-closure-index',
      expectedClosureCoordinate: 55,
      expectedClosureIndex: 1,
      closureDistanceFromExactAdmission: 12,
      returnsToBase: true,
    });
    expect(secondClosureRejection?.family).toEqual(secondClosureFamily);
    expect(secondClosureRejection?.theoremIds).toContain(
      'GamutBraid.indexed_interfere_closure_distance_positive_for_later'
    );
    expect(secondClosureRejection?.theoremIds).toContain(
      'GamutBraid.indexed_interfere_port_later_closure_family_rejected_by_exact_admission'
    );
    expect(secondClosureRejection?.theoremIds).toContain(
      'GamutBraid.indexed_interfere_port_second_closure_family_rejected_by_exact_admission'
    );
    expect(thirdClosureFamily).toMatchObject({
      source: 'GnosisIndexedInterferePortClosureFamilyCertificate',
      portAction: 'interfere',
      portCoordinate: 43,
      closureCoordinate: 79,
      closureIndex: 3,
      closureStride: 12,
      closureStepsAfterExactAdmission: 2,
      closureDistanceFromExactAdmission: 24,
      closureAxisPressureWeightedTotal: 6042,
      closureAxisPressureGrowthFromInterferencePort: 3456,
      closureAxisPressureGrowthFromExactClosure: 2304,
      orderedAfterExactAdmission: true,
      exactAdmissionCompatible: false,
      terminal: true,
      returnsToBase: true,
    });
    expect(thirdClosureFamily?.theoremIds).toContain(
      'GamutBraid.exact_indexed_interfere_port_closure_family_member_certified'
    );
    expect(thirdClosureFamily?.theoremIds).toContain(
      'GamutBraid.exact_indexed_interfere_port_third_closure_family_distance'
    );
    expect(thirdClosureFamily?.theoremIds).toContain(
      'GamutBraid.exact_indexed_interfere_port_third_closure_axis_pressure_total'
    );
    expect(thirdClosureRejection).toMatchObject({
      source: 'GnosisIndexedInterferePortClosureFamilyRejectionCertificate',
      rejectedBy: 'GnosisIndexedInterferePortAdmissionCertificate',
      reason: 'wrong-closure-index',
      expectedClosureCoordinate: 55,
      expectedClosureIndex: 1,
      closureDistanceFromExactAdmission: 24,
      returnsToBase: true,
    });
    expect(thirdClosureRejection?.family).toEqual(thirdClosureFamily);
    expect(thirdClosureRejection?.theoremIds).toContain(
      'GamutBraid.indexed_interfere_closure_distance_positive_for_later'
    );
    expect(thirdClosureRejection?.theoremIds).toContain(
      'GamutBraid.indexed_interfere_port_later_closure_family_rejected_by_exact_admission'
    );
    expect(closureFamilyAudit).toMatchObject({
      source: 'GnosisIndexedInterferePortClosureFamilyAuditCertificate',
      portAction: 'interfere',
      portCoordinate: 43,
      closureStride: 12,
      maxClosureIndex: 5,
      closureIndices: [1, 2, 3, 4, 5],
      pressureGrowthBlock: 1152,
      closureFamilyPressureTotal: 30210,
      pressureUnit: 'SPa',
      exactAdmissionClosureIndex: 1,
      exactAdmissionClosureCoordinate: 55,
      returnsToBase: true,
    });
    expect(
      closureFamilyAudit?.members.map((member) => ({
        closureIndex: member.closureIndex,
        closureCoordinate: member.closureCoordinate,
        closureDistanceFromExactAdmission:
          member.closureDistanceFromExactAdmission,
        closureAxisPressureWeightedTotal:
          member.closureAxisPressureWeightedTotal,
        closureAxisPressureGrowthFromInterferencePort:
          member.closureAxisPressureGrowthFromInterferencePort,
        closureAxisPressureGrowthFromExactClosure:
          member.closureAxisPressureGrowthFromExactClosure,
        exactAdmissionCompatible: member.exactAdmissionCompatible,
      }))
    ).toEqual([
      {
        closureIndex: 1,
        closureCoordinate: 55,
        closureDistanceFromExactAdmission: 0,
        closureAxisPressureWeightedTotal: 3738,
        closureAxisPressureGrowthFromInterferencePort: 1152,
        closureAxisPressureGrowthFromExactClosure: 0,
        exactAdmissionCompatible: true,
      },
      {
        closureIndex: 2,
        closureCoordinate: 67,
        closureDistanceFromExactAdmission: 12,
        closureAxisPressureWeightedTotal: 4890,
        closureAxisPressureGrowthFromInterferencePort: 2304,
        closureAxisPressureGrowthFromExactClosure: 1152,
        exactAdmissionCompatible: false,
      },
      {
        closureIndex: 3,
        closureCoordinate: 79,
        closureDistanceFromExactAdmission: 24,
        closureAxisPressureWeightedTotal: 6042,
        closureAxisPressureGrowthFromInterferencePort: 3456,
        closureAxisPressureGrowthFromExactClosure: 2304,
        exactAdmissionCompatible: false,
      },
      {
        closureIndex: 4,
        closureCoordinate: 91,
        closureDistanceFromExactAdmission: 36,
        closureAxisPressureWeightedTotal: 7194,
        closureAxisPressureGrowthFromInterferencePort: 4608,
        closureAxisPressureGrowthFromExactClosure: 3456,
        exactAdmissionCompatible: false,
      },
      {
        closureIndex: 5,
        closureCoordinate: 103,
        closureDistanceFromExactAdmission: 48,
        closureAxisPressureWeightedTotal: 8346,
        closureAxisPressureGrowthFromInterferencePort: 5760,
        closureAxisPressureGrowthFromExactClosure: 4608,
        exactAdmissionCompatible: false,
      },
    ]);
    expect(closureFamilyAudit?.theoremIds).toContain(
      'GamutBraid.indexed_interfere_closure_family_audit_trace_members_certified'
    );
    expect(closureFamilyAudit?.theoremIds).toContain(
      'GamutBraid.indexed_interfere_closure_family_audit_trace_pressure_total_sum'
    );
    expect(closureFamilyAudit?.theoremIds).toContain(
      'GamutBraid.bounded_indexed_interfere_closure_family_audit_trace_certified'
    );
    expect(closureFamilyAudit?.theoremIds).toContain(
      'GamutBraid.exact_bounded_indexed_interfere_closure_family_audit_trace'
    );
    expect(closureFamilyAudit?.theoremIds).toContain(
      'GamutBraid.exact_bounded_indexed_interfere_closure_family_audit_pressure_total'
    );
    expect(sparseClosureFamilyAudit?.closureIndices).toEqual([1, 4, 5]);
    expect(sparseClosureFamilyAudit?.closureFamilyPressureTotal).toBe(19278);
    expect(sparseClosureFamilyAudit?.pressureUnit).toBe('SPa');
    expect(
      sparseClosureFamilyAudit?.members.map(
        (member) => member.closureAxisPressureWeightedTotal
      )
    ).toEqual([3738, 7194, 8346]);
    expect(sparseClosureFamilyAudit?.theoremIds).toContain(
      'GamutBraid.indexed_interfere_closure_family_audit_trace_members_certified'
    );
    expect(sparseClosureFamilyAudit?.theoremIds).not.toContain(
      'GamutBraid.exact_bounded_indexed_interfere_closure_family_audit_trace'
    );
    expect(indexedAdmission).toMatchObject({
      source: 'GnosisIndexedInterferePortAdmissionCertificate',
      portAction: 'interfere',
      portCoordinate: 43,
      closureCoordinate: 55,
      closureIndex: 1,
      closureStride: 12,
      terminal: true,
      returnsToBase: true,
    });
    expect(indexedAdmission?.closureFamily).toEqual(firstClosureFamily);
    expect(indexedAdmission?.transition).toEqual(certificate);
    expect(indexedAdmission?.theoremIds).toContain(
      'GamutBraid.exact_indexed_interfere_port_closure_family_certified'
    );
    expect(indexedAdmission?.theoremIds).toContain(
      'GamutBraid.exact_indexed_interfere_port_admission_certified'
    );
    expect(indexedAdmission?.theoremIds).toContain(
      'GamutBraid.indexed_interfere_port_admission_forces_exact_coordinates'
    );
    expect(interferePort).toMatchObject({
      source: 'GnosisUringInterferePortCertificate',
      host: 'gnosis-uring',
      portAction: 'interfere',
      portCoordinate: 43,
      closureCoordinate: 55,
      terminal: true,
      returnsToBase: true,
    });
    expect(interferePort?.indexedAdmission).toEqual(indexedAdmission);
    expect(interferePort?.closureFamilyAudit).toEqual(closureFamilyAudit);
    expect(interferePort?.transition).toEqual(certificate);
    expect(interferePort?.theoremIds).toContain(
      'GamutBraid.exact_interfere_port_certified'
    );
    expect(interferePort?.theoremIds).toContain(
      'GamutBraid.exact_indexed_interfere_port_closure_family_member_certified'
    );
    expect(interferePort?.theoremIds).toContain(
      'GamutBraid.indexed_interfere_port_later_closure_family_rejected_by_exact_admission'
    );
    expect(interferePort?.theoremIds).toContain(
      'GamutBraid.exact_indexed_interfere_port_second_closure_family_certified'
    );
    expect(interferePort?.theoremIds).toContain(
      'GamutBraid.indexed_interfere_port_second_closure_family_rejected_by_exact_admission'
    );
    expect(stride).toMatchObject({
      source: 'GnosisInterferenceStrideCertificate',
      portAction: 'interfere',
      anchorCoordinate: 43,
      closureCoordinate: 55,
      naturalCoordinates: [10, 12, 17, 22],
      strideTrace: [33, 31, 26, 21],
      externalToNaturalStack: true,
      terminal: true,
      returnsToBase: true,
    });
    expect(stride?.braidPacket).toEqual(packet);
    expect(stride?.theoremIds).toContain(
      'GamutBraid.exact_interference_anchor_packet_certified'
    );
    expect(pressureWeightedStride).toMatchObject({
      source: 'GnosisPressureWeightedInterferenceStrideCertificate',
      portAction: 'interfere',
      anchorCoordinate: 43,
      closureCoordinate: 55,
      strideTrace: [33, 31, 26, 21],
      leakageTrace: [0, 18, 78],
      leakageTotal: 96,
      weightedPressureTrace: [0, 558, 2028],
      weightedPressureTotal: 2586,
      pressureUnit: 'SPa',
      weighting: 'stride-leakage-zip',
      returnsToBase: true,
    });
    expect(pressureWeightedStride?.interferenceStride).toEqual(stride);
    expect(pressureWeightedStride?.theoremIds).toContain(
      'GamutBraid.exact_pressure_weighted_interference_stride_trace'
    );
    expect(pressureWeightedStride?.theoremIds).toContain(
      'GamutBraid.exact_pressure_weighted_interference_total'
    );
    expect(pressureWeightedStride?.theoremIds).toContain(
      'GamutBraid.exact_pressure_weighted_interference_anchor_packet_certified'
    );
    expect(pressureWeightedStride?.theoremIds).toContain(
      'GamutBraid.exact_pressure_weighted_interfere_port_closure_family_first_member'
    );
    expect(pressureWeightedStride?.theoremIds).toContain(
      'GamutBraid.pressure_weighted_interfere_port_closure_family_first_member_forces_exact_coordinates'
    );
    expect(pressureWeightedStride?.theoremIds).toContain(
      'GamutBraid.exact_pressure_weighted_interfere_port_closure_family_member_certified'
    );
    expect(pressureWeightedStride?.theoremIds).toContain(
      'GamutBraid.pressure_weighted_interfere_port_later_closure_family_rejected_by_exact_admission'
    );
    expect(pressureWeightedStride?.theoremIds).toContain(
      'GamutBraid.exact_pressure_weighted_interfere_port_second_closure_family_certified'
    );
    expect(createExtendedAeonTransitionCertificate(malformedPacket)).toBe(null);
    expect(createExtendedAeonTransitionCertificate(null)).toBe(null);
    expect(
      createIndexedInterferePortClosureFamilyCertificate(certificate, 44, 2)
    ).toBe(null);
    expect(
      createIndexedInterferePortClosureFamilyCertificate(certificate, 43, 6)
    ).toBe(null);
    expect(
      createIndexedInterferePortClosureFamilyAuditCertificate(
        certificate,
        [1, 6]
      )
    ).toBe(null);
    expect(createIndexedInterferePortClosureFamilyAuditCertificate(null)).toBe(
      null
    );
    expect(
      createIndexedInterferePortClosureFamilyRejectionCertificate(
        firstClosureFamily
      )
    ).toBe(null);
    expect(createIndexedInterferePortAdmissionCertificate(null)).toBe(null);
    expect(
      createIndexedInterferePortAdmissionCertificate(certificate, 44, 56)
    ).toBe(null);
    expect(
      createIndexedInterferePortAdmissionCertificate(certificate, 43, 67, 2)
    ).toBe(null);
    expect(createGnosisUringInterferePortCertificate(null)).toBe(null);
    expect(createInterferenceStrideCertificate(malformedPacket)).toBe(null);
    expect(createInterferenceStrideCertificate(null)).toBe(null);
    expect(createPressureWeightedInterferenceStrideCertificate(null)).toBe(
      null
    );
  });

  it('mirrors the finite gnostic lens geometry certificate', () => {
    const packet = createBraidPacketCertificate(3);
    const lens = createGnosticLensGeometryCertificate(packet);
    const malformedPacket =
      packet === null
        ? null
        : ({
            ...packet,
            portCoordinates: [10, 12, 17, 23],
          } as unknown as BraidPacketCertificate);

    expect(lens).toMatchObject({
      source: 'GnosisGnosticLensGeometryCertificate',
      sphericalSaturationTrace: [10, 30, 90],
      sphericalInformationMassTrace: [120, 360, 1080],
      sphericalRotationTrace: [120, 360, 1080],
      sphericalWrapTrace: [1, 1, 3],
      packedGeometryTotalMass: 1560,
      packedGeometryBucketWeight: 130,
      highLatitudePortCoordinates: [17, 22],
      lensObservationCoordinates: [12, 30, 90, 43],
      lensObservationShapes: ['plane', 'sphere', 'torus', 'singularAxis'],
      portLensObservationCoordinates: [10, 12, 17, 22, 43],
      portLensObservationShapes: [
        'plane',
        'plane',
        'torus',
        'torus',
        'singularAxis',
      ],
      nestedGeometry: {
        enclosureShape: 'sphere',
        braidShape: 'torus',
        enclosureRotationDegrees: 360,
        braidRotationDegrees: 1080,
        braidWrapCount: 3,
        brownRotationIsThreePinkRotations: true,
      },
      twistedDoubleKeystone: {
        bucket: 12,
        halfCarrier: 11,
        doubleKeystone: 22,
        perfectDoubleCover: 24,
        coverDeficit: 2,
        foldSwerve: 17,
        crossingCount: 5,
        leakagePressure: 96,
        flatReferencePressure: 0,
        pressureExceedsFlatReference: true,
      },
      toroidalAxis: {
        coordinate: 43,
        externalToNaturalStack: true,
        terminal: true,
        closureCoordinate: 55,
        closureStride: 12,
      },
      packedLensRoleTrace: ['coreFilament', 'innerRing', 'outerBraidedShell'],
      packedLensMassTrace: [120, 360, 1080],
      packedLensShapeTrace: ['plane', 'sphere', 'torus'],
      centerlineGamutTrace: [10, 12, 17, 22, 43],
      verdict: {
        localResolution: 12,
        executionPathShape: 'torus',
        informationContainerShape: 'sphere',
        terminalInterferenceShape: 'singularAxis',
        totalPackedMass: 1560,
        centerlineTrace: [10, 12, 17, 22, 43],
        finiteSelfReferenceShape: 'selfReferenceKnot',
      },
      returnsToBase: true,
    });
    expect(lens?.braidPacket).toEqual(packet);
    expect(lens?.harmonicShortcuts).toEqual([
      {
        shell: 'pink',
        fullRotationDegrees: 360,
        sampledDegrees: 180,
        mirroredDegrees: 180,
        quotientFactor: 2,
        fullQuadraticWork: 129600,
        quotientQuadraticWork: 64800,
        halvesQuadraticWork: true,
        reducesWork: true,
      },
      {
        shell: 'brown',
        fullRotationDegrees: 1080,
        sampledDegrees: 540,
        mirroredDegrees: 540,
        quotientFactor: 2,
        fullQuadraticWork: 1166400,
        quotientQuadraticWork: 583200,
        halvesQuadraticWork: true,
        reducesWork: true,
      },
    ]);
    expect(lensShapeForPortCoordinate(10)).toBe('plane');
    expect(lensShapeForPortCoordinate(12)).toBe('plane');
    expect(lensShapeForPortCoordinate(17)).toBe('torus');
    expect(lensShapeForPortCoordinate(22)).toBe('torus');
    expect(lensShapeForPortCoordinate(43)).toBe('singularAxis');
    expect(lensShapeForPortCoordinate(55)).toBe(null);
    expect(lens?.theoremIds).toContain(
      'GamutBraid.exact_gnostic_lens_geometry_certified'
    );
    expect(lens?.theoremIds).toContain(
      'GamutBraid.exact_gnostic_lens_verdict_certified'
    );
    expect(lens?.theoremIds).toContain(
      'GamutBraid.exact_interference_closure_stride_is_bucket'
    );
    expect(createGnosticLensGeometryCertificate(malformedPacket)).toBe(null);
    expect(createGnosticLensGeometryCertificate(null)).toBe(null);
  });

  it('exports the proof-backed closed-loop runtime metadata checksum', () => {
    const packet = createBraidPacketCertificate(3);
    const lens = createGnosticLensGeometryCertificate(packet);
    const stride = createInterferenceStrideCertificate(packet);
    const pressureWeighted =
      createPressureWeightedInterferenceStrideCertificate(stride);
    const metadata = createClosedLoopRuntimeMetadataCertificate(
      lens,
      pressureWeighted
    );
    const mismatchedPacket = createBraidPacketCertificate(4);
    const mismatchedPressureWeighted =
      createPressureWeightedInterferenceStrideCertificate(
        createInterferenceStrideCertificate(mismatchedPacket)
      );
    const malformedLens =
      lens === null
        ? null
        : ({
            ...lens,
            packedGeometryBucketWeight: 131,
          } as unknown as GnosticLensGeometryCertificate);
    const malformedPressureWeighted =
      pressureWeighted === null
        ? null
        : ({
            ...pressureWeighted,
            weightedPressureTotal: 2585,
          } as unknown as PressureWeightedInterferenceStrideCertificate);

    expect(metadata).toMatchObject({
      source: 'GnosisClosedLoopRuntimeMetadataCertificate',
      cycle: 3,
      closedTrace: [10, 12, 17, 22, 43, 55, 10],
      centerlineTrace: [10, 12, 17, 22, 43],
      representativeTrace: [180, 180],
      reconstructedTrace: [360, 1080],
      axisCoordinate: 43,
      closureCoordinate: 55,
      totalMass: 1560,
      totalLeakage: 96,
      pressureWeightedTotal: 2586,
      bucketRatio: 130,
      checksum: 6439,
      returnsToBase: true,
    });
    expect(metadata?.gnosticLens).toEqual(lens);
    expect(metadata?.pressureWeightedInterference).toEqual(pressureWeighted);
    expect(
      metadata === null ? 0 : closedLoopRuntimeMetadataChecksum(metadata)
    ).toBe(6439);
    expect(metadata?.theoremIds).toContain(
      'GamutBraid.exact_closed_loop_certified'
    );
    expect(metadata?.theoremIds).toContain(
      'GamutBraid.exact_closed_loop_runtime_metadata_certified'
    );
    expect(metadata?.theoremIds).toContain(
      'GamutBraid.exact_cyclic_compiler_reconstructed_trace'
    );
    expect(metadata?.theoremIds).toContain(
      'GamutBraid.exact_pressure_weighted_closed_loop_workload_certified'
    );
    expect(
      createClosedLoopRuntimeMetadataCertificate(
        lens,
        mismatchedPressureWeighted
      )
    ).toBe(null);
    expect(
      createClosedLoopRuntimeMetadataCertificate(
        malformedLens,
        pressureWeighted
      )
    ).toBe(null);
    expect(
      createClosedLoopRuntimeMetadataCertificate(
        lens,
        malformedPressureWeighted
      )
    ).toBe(null);
    expect(
      createClosedLoopRuntimeMetadataCertificate(null, pressureWeighted)
    ).toBe(null);
  });

  it('certifies pressure-weighted closed-loop benchmark report identity', () => {
    const metadata = createClosedLoopRuntimeMetadataCertificate();
    const report =
      createPressureWeightedClosedLoopBenchmarkReportCertificate(metadata);
    const malformedMetadata =
      metadata === null
        ? null
        : ({
            ...metadata,
            pressureWeightedTotal: 2585,
          } as unknown as ClosedLoopRuntimeMetadataCertificate);

    expect(report).toMatchObject({
      source: 'GnosisPressureWeightedClosedLoopBenchmarkReportCertificate',
      topologyCoordinates: [10, 12, 17, 22, 43],
      gateCoordinates: [10, 12, 17, 22],
      decompositionCoordinates: [7, 17, 22],
      primeDoubleCrossingCount: 5,
      interferePortCoordinate: 43,
      interfereClosureCoordinate: 55,
      interferenceStrideTrace: [33, 31, 26, 21],
      leakagePressureTrace: [0, 18, 78],
      pressureWeightedInterferenceStrideTrace: [0, 558, 2028],
      pressureWeightedInterferenceTotal: 2586,
      pressureUnit: 'SPa',
      returnsToBase: true,
    });
    expect(pressureWeightedBenchmarkTopologyCoordinatesFor(43)).toEqual([
      10, 12, 17, 22, 43,
    ]);
    expect(pressureWeightedBenchmarkInterferenceStrideTraceFor(43)).toEqual([
      33, 31, 26, 21,
    ]);
    expect(pressureWeightedBenchmarkWeightedStrideTraceFor(43)).toEqual([
      0, 558, 2028,
    ]);
    expect(pressureWeightedBenchmarkTotalFor(43)).toBe(2586);
    expect(report?.representativeWorkloadTheoremIds).toEqual([
      'GamutBraid.exact_closed_loop_weight_accumulation_representative_shortcut',
      'GamutBraid.exact_closed_loop_repeated_unit_workload_certified',
      'GamutBraid.exact_pressure_weighted_closed_loop_workload_certified',
    ]);
    expect(report?.theoremIds).toContain(
      'GamutBraid.exact_pressure_weighted_closed_loop_benchmark_report_family_certified'
    );
    expect(report?.theoremIds).toContain(
      'GamutBraid.exact_pressure_weighted_closed_loop_benchmark_report_certified'
    );
    expect(report?.closedLoopRuntimeMetadata).toEqual(metadata);
    expect(
      createPressureWeightedClosedLoopBenchmarkReportCertificate(
        malformedMetadata
      )
    ).toBe(null);
    expect(
      createPressureWeightedClosedLoopBenchmarkReportCertificate(null)
    ).toBe(null);
  });

  it('certifies the Wankel runtime fast path over the pressure execution certificate', () => {
    const metadata = createClosedLoopRuntimeMetadataCertificate();
    const report =
      createPressureWeightedClosedLoopBenchmarkReportCertificate(metadata);
    const pressurePath = createPressureCertifiedExecutionPathCertificate({
      pressureWeightedClosedLoopBenchmarkReport: report,
    });
    const certificate = createWankelRuntimeCertificate({
      closedLoopRuntimeMetadata: metadata,
      pressureWeightedClosedLoopBenchmarkReport: report,
      pressureCertifiedExecutionPath: pressurePath,
    });
    const mismatchedReport =
      createPressureWeightedClosedLoopBenchmarkReportCertificate();
    const malformedCertificate =
      certificate === null
        ? null
        : ({
            ...certificate,
            deadlinePulseHops: 256,
          } as unknown as NonNullable<typeof certificate>);

    expect(certificate).toMatchObject({
      source: 'GnosisWankelRuntimeCertificate',
      runtime: 'gnosis',
      host: 'gnosis-uring',
      aeonCells: WANKEL_AEON_CELLS,
      lifecycleTrace: WANKEL_FIVE_FORCE_LIFECYCLE_TRACE,
      deadlinePulseHops: WANKEL_RUNTIME_DEADLINE_PULSE_HOPS,
      returnsToBase: true,
    });
    expect(certificate?.closedLoopRuntimeMetadata).toBe(metadata);
    expect(certificate?.pressureWeightedClosedLoopBenchmarkReport).toBe(
      report
    );
    expect(certificate?.pressureCertifiedExecutionPath).toBe(pressurePath);
    expect(certificate?.theoremIds).toEqual(
      WANKEL_RUNTIME_CERTIFICATE_THEOREM_IDS
    );
    expect(certificate?.theoremIds).toContain(
      'WankelEngineTheorem.wankel_schedules_interference'
    );
    expect(certificate?.theoremIds).toContain(
      'RuntimeCertificate.wankel_fifth_force_mechanism_has_runtime_certificate'
    );
    expect(certificate?.theoremIds).toContain(
      'RuntimeCertificate.wankel_fast_path_cost_dominates_baseline'
    );
    expect(certificate?.theoremIds).toContain(
      'RuntimeCertificate.wankel_aggregate_fast_path_cost_dominates_baseline'
    );
    expect(certificate?.theoremIds).toContain(
      'RuntimeCertificate.wankel_robust_fast_path_estimator_dominates_baseline'
    );
    expect(selectWankelRuntimeFastPath(certificate)).toMatchObject({
      selected: true,
      reason: 'wankel runtime fast path selected',
      deadlinePulseHops: WANKEL_RUNTIME_DEADLINE_PULSE_HOPS,
      lifecycleTrace: WANKEL_FIVE_FORCE_LIFECYCLE_TRACE,
    });
    expect(selectWankelRuntimeFastPath(null)).toMatchObject({
      selected: false,
      reason: 'wankel runtime certificate is absent',
      deadlinePulseHops: null,
    });
    expect(selectWankelRuntimeFastPath(malformedCertificate)).toMatchObject({
      selected: false,
      reason: 'wankel runtime certificate is inexact',
    });
    expect(
      createWankelRuntimeCertificate({
        closedLoopRuntimeMetadata: metadata,
        pressureWeightedClosedLoopBenchmarkReport: mismatchedReport,
      })
    ).toBe(null);
    expect(createWankelRuntimeCertificate({ closedLoopRuntimeMetadata: null }))
      .toBe(null);
    expect(
      certifyWankelRuntimeFastPathEquivalence(certificate, 6439, 6439)
        .equivalent
    ).toBe(true);
    expect(
      certifyWankelRuntimeFastPathEquivalence(certificate, 6439, 6440)
        .equivalent
    ).toBe(false);
    expect(
      certifyWankelRuntimeFastPathCostDominance(
        certificate,
        6439,
        6439,
        {
          selectedCost: 88,
          baselineCost: 144,
          unit: 'nsPerIter',
        }
      )
    ).toMatchObject({
      dominates: true,
      reason: 'wankel runtime fast path cost dominance certified',
      theoremId: 'RuntimeCertificate.wankel_fast_path_cost_dominates_baseline',
    });
    expect(
      certifyWankelRuntimeFastPathCostDominance(
        certificate,
        6439,
        6439,
        {
          selectedCost: 144,
          baselineCost: 88,
          unit: 'nsPerIter',
        }
      )
    ).toMatchObject({
      dominates: false,
      reason: 'wankel runtime fast path is not faster than baseline',
    });
    expect(
      certifyWankelRuntimeFastPathAggregateCostDominance(
        certificate,
        6439,
        6439,
        {
          sampleCount: 7,
          selectedTotalCost: 616,
          baselineTotalCost: 1008,
          unit: 'nsPerIter',
        }
      )
    ).toMatchObject({
      dominates: true,
      reason: 'wankel runtime aggregate fast path cost dominance certified',
      theoremId:
        'RuntimeCertificate.wankel_aggregate_fast_path_cost_dominates_baseline',
    });
    expect(
      certifyWankelRuntimeFastPathAggregateCostDominance(
        certificate,
        6439,
        6439,
        {
          sampleCount: 0,
          selectedTotalCost: 616,
          baselineTotalCost: 1008,
          unit: 'nsPerIter',
        }
      )
    ).toMatchObject({
      dominates: false,
      reason:
        'wankel runtime aggregate cost witness sample count must be a positive integer',
    });
    expect(
      certifyWankelRuntimeFastPathRobustEstimatorDominance(
        certificate,
        6439,
        6439,
        {
          sampleCount: 7,
          minimumSampleCount: 3,
          selectedTotalCost: 616,
          baselineTotalCost: 1008,
          selectedRobustCost: 88,
          baselineRobustCost: 144,
          observedJitter: 5,
          jitterBudget: 8,
          outlierBudget: 1,
          unit: 'nsPerIter',
        }
      )
    ).toMatchObject({
      dominates: true,
      reason: 'wankel runtime robust estimator cost dominance certified',
      theoremId:
        'RuntimeCertificate.wankel_robust_fast_path_estimator_dominates_baseline',
    });
    expect(
      certifyWankelRuntimeFastPathRobustEstimatorDominance(
        certificate,
        6439,
        6439,
        {
          sampleCount: 2,
          minimumSampleCount: 3,
          selectedTotalCost: 616,
          baselineTotalCost: 1008,
          selectedRobustCost: 88,
          baselineRobustCost: 144,
          observedJitter: 5,
          jitterBudget: 8,
          outlierBudget: 0,
          unit: 'nsPerIter',
        }
      )
    ).toMatchObject({
      dominates: false,
      reason:
        'wankel runtime robust estimator witness requires minimum sample coverage',
    });
  });

  it('certifies the complete pressure execution path as one theorem-backed object', () => {
    const pressurePath = createPressureCertifiedExecutionPathCertificate();
    const malformedPathBenchmark =
      pressurePath === null
        ? null
        : ({
            ...pressurePath.pressureWeightedClosedLoopBenchmarkReport,
            pressureWeightedInterferenceTotal: 2587,
          } as unknown as PressureCertifiedExecutionPathCertificate['pressureWeightedClosedLoopBenchmarkReport']);

    expect(pressurePath).toMatchObject({
      source: 'GnosisPressureCertifiedExecutionPathCertificate',
      certificateVersion: 'pressure-certified-execution-path/v1',
      pressureIdentity: EXACT_PRESSURE_CERTIFIED_EXECUTION_PATH_IDENTITY,
      runtime: 'gnosis',
      host: 'gnosis-uring',
      topologyCoordinates: [10, 12, 17, 22, 43],
      gateCoordinates: [10, 12, 17, 22],
      decompositionCoordinates: [7, 17, 22],
      primeDoubleCrossingCount: 5,
      interferePort: {
        action: 'interfere',
        coordinate: 43,
        closureCoordinate: 55,
        terminal: true,
      },
      pressureWeightedInterference: {
        strideTrace: [33, 31, 26, 21],
        leakagePressureTrace: [0, 18, 78],
        pressureWeightedStrideTrace: [0, 558, 2028],
        pressureWeightedTotal: 2586,
        pressureUnit: 'SPa',
      },
      informationMassTotal: 1560,
      leakageTotal: 96,
      metrology: {
        informationMassTotal: 1560,
        leakageTotal: 96,
        massUnit: 'gG',
        pressureUnit: 'SPa',
      },
      returnsToBase: true,
    });
    expect(pressurePath?.closureFamilyAudit).toMatchObject({
      closureFamilyPressureTotal: 30210,
      pressureUnit: 'SPa',
      closureIndices: [1, 2, 3, 4, 5],
    });
    expect(pressurePath?.pressureWeightedClosedLoopBenchmarkReport).toEqual(
      createPressureWeightedClosedLoopBenchmarkReportCertificate()
    );
    expect(pressurePath?.theoremIds).toContain(
      'GamutBraid.exact_pressure_certified_execution_path'
    );
    expect(pressurePath?.theoremIds).toContain(
      'GamutBraid.pressure_certified_execution_path_closure_family_checksum'
    );
    expect(pressurePath?.theoremIds).toContain(
      'StatisticalMechanicsAccounting.exact_indexed_pressure_selector_stable_tie_contract'
    );
    const stableTieAnchors =
      PRESSURE_CERTIFIED_EXECUTION_PATH_STABLE_TIE_ANCHOR_THEOREM_IDS;
    for (const stableTieAnchor of stableTieAnchors) {
      expect(pressurePath?.theoremIds).toContain(stableTieAnchor);
    }
    expect(
      createPressureCertifiedExecutionPathCertificate({
        pressureWeightedClosedLoopBenchmarkReport: malformedPathBenchmark,
      })
    ).toBe(null);
    expect(
      createPressureCertifiedExecutionPathCertificate({
        closureFamilyAudit:
          pressurePath === null
            ? null
            : ({
                ...pressurePath.closureFamilyAudit,
                closureFamilyPressureTotal: 30211,
              } as unknown as PressureCertifiedExecutionPathCertificate['closureFamilyAudit']),
      })
    ).toBe(null);
  });

  it('certifies pressure gradient and fluidic allocation from metrology', () => {
    const accounting = createStatisticalMechanicsAccounting(
      createClosedArchitectureCertificate()
    );
    const metrology = createGnosticMetrologyProfile(accounting);
    const pressureGradient = createPressureGradientCertificate(metrology);
    const acceptedPlan = createGnosticRuntimeAllocationPlan(1560, metrology);
    const rejectedPlan = createGnosticRuntimeAllocationPlan(1559, metrology);
    const fluidicBridge = createCalibratedFluidicPressureBridge(2, metrology);
    const routingCertificate = createFluidicRoutingAdmissibilityCertificate(
      2,
      1560,
      metrology
    );

    expect(pressureGradient).toMatchObject({
      bucketSize: 12,
      bucketCount: 8,
      totalSievePressure: 96,
      bucketProduct: 96,
      reserveSievePressure: 96,
    });
    expect(acceptedPlan?.admitted).toBe(true);
    expect(acceptedPlan?.reserveBrownBufferMass).toBe(1080);
    expect(acceptedPlan?.reserveSievePressure).toBe(96);
    expect(rejectedPlan?.admitted).toBe(false);
    expect(rejectedPlan?.reason).toBe('checksum-mismatch');
    expect(fluidicBridge?.pressureTracePa).toEqual([0, 36, 156]);
    expect(fluidicBridge?.totalPhysicalPressurePa).toBe(192);
    expect(fluidicBridge?.brownPhysicalPressurePa).toBe(156);
    expect(routingCertificate).toMatchObject({
      observedMass: 1560,
      expectedMass: 1560,
      checksumMatches: true,
      pressureRankedBands: ['brown', 'pink', 'white'],
      dominantBand: 'brown',
      dominantSievePressure: 78,
      dominantPhysicalPressurePa: 156,
      pinkPhysicalPressurePa: 36,
      totalPhysicalPressurePa: 192,
      preservesGradient: true,
      admitted: true,
      returnsToBase: true,
    });
    expect(routingCertificate?.allocationPlan).toEqual(acceptedPlan);
    expect(routingCertificate?.fluidicBridge).toEqual(fluidicBridge);
    expect(
      createFluidicRoutingAdmissibilityCertificate(2, 1559, metrology)
    ).toBe(null);
    expect(
      createFluidicRoutingAdmissibilityCertificate(0, 1560, metrology)
    ).toBe(null);
    expect(createCalibratedFluidicPressureBridge(0, metrology)).toBe(null);
    expect(createPressureGradientCertificate(null)).toBe(null);
  });

  it('requires exact witness constraints before issuing accounting', () => {
    const braidCertificate = createBraidFoldCertificate(exactCandidate);
    const decompositionCertificate = createDecompositionTripleCertificate(
      {
        cycle: 3,
        coordinates: [7, 17, 22],
      },
      braidCertificate
    );
    const malformedBraidCertificate: BraidFoldCertificate | null =
      braidCertificate === null
        ? null
        : ({
            ...braidCertificate,
            saturationTrace: [10, 30, 91],
          } as unknown as BraidFoldCertificate);

    expect(
      createStatisticalMechanicsAccounting(decompositionCertificate)
        ?.leakageTrace
    ).toEqual([0, 18, 78]);
    expect(createStatisticalMechanicsAccounting(null)).toBe(null);
    expect(
      createStatisticalMechanicsAccounting(malformedBraidCertificate)
    ).toBe(null);
  });

  it('requires certified and linear results to stay semantically equivalent', () => {
    const certificate = createBraidFoldCertificate(exactCandidate);
    const linearResult = { payload: 122, path: [10, 12, 17, 22] };
    const certifiedResult = { payload: 122, path: [10, 12, 17, 22] };
    const mismatchResult = { payload: 123, path: [10, 17, 22] };
    const equivalent = (
      left: typeof linearResult,
      right: typeof linearResult
    ) => JSON.stringify(left) === JSON.stringify(right);

    expect(
      certifyBraidFastPathEquivalence(
        certificate,
        linearResult,
        certifiedResult,
        equivalent
      ).equivalent
    ).toBe(true);
    expect(
      certifyBraidFastPathEquivalence(
        certificate,
        linearResult,
        mismatchResult,
        equivalent
      ).equivalent
    ).toBe(false);
    expect(
      certifyBraidFastPathEquivalence(
        null,
        linearResult,
        certifiedResult,
        equivalent
      ).decision.selected
    ).toBe(false);
  });

  it('lifts mesh-policy admissions to canonical ternary verdicts with a real defer middle', () => {
    // ── Site 1: saturation lifecycle admission (observeSaturation) ──────
    // On-gate saturations commit to a layer → Accept.
    const onGate = observeSaturation(30);
    expect(onGate.verdict).toBe(Verdict.Accept);
    expect(isAccept(onGate.verdict)).toBe(true);
    expect(onGate.admitted).toBe(isAccept(onGate.verdict));

    // Saturation 12 sits strictly between gates 10 and 30: qualified but not
    // yet scheduled to a layer → Abstain (the neutral middle a bool drops).
    const deferred = observeSaturation(12);
    expect(deferred.verdict).toBe(Verdict.Abstain);
    expect(isDefer(deferred.verdict)).toBe(true);
    // Lossy collapse is preserved: the legacy boolean still reads false.
    expect(deferred.admitted).toBe(false);
    expect(deferred.admitted).toBe(isAccept(deferred.verdict));

    // Outside the keystone band (or non-finite) is disqualified → Decline.
    const declinedLow = observeSaturation(5);
    expect(declinedLow.verdict).toBe(Verdict.Decline);
    expect(isReject(declinedLow.verdict)).toBe(true);
    expect(observeSaturation(Number.NaN).verdict).toBe(Verdict.Decline);
    // Abstain and Decline both collapse to the same false — the bool cannot
    // tell deferral from disqualification; the verdict can.
    expect(deferred.admitted).toBe(declinedLow.admitted);
    expect(deferred.verdict).not.toBe(declinedLow.verdict);

    // ── Site 2: braid-fold audit (auditBraidFoldCandidate) ──────────────
    const accepted = auditBraidFoldCandidate(exactCandidate);
    expect(accepted.verdict).toBe(Verdict.Accept);
    expect(accepted.accepted).toBe(isAccept(accepted.verdict));

    // Right canonical shape, readiness flag not yet asserted → Abstain.
    const unscheduled = auditBraidFoldCandidate({
      ...exactCandidate,
      baseReady: false,
    });
    expect(unscheduled.verdict).toBe(Verdict.Abstain);
    expect(isDefer(unscheduled.verdict)).toBe(true);
    expect(unscheduled.accepted).toBe(false);

    // Wrong shape (coordinates can never be the braid fold) → Decline.
    const wrongShape = auditBraidFoldCandidate({
      ...exactCandidate,
      gateCoordinates: [12, 17, 22],
    });
    expect(wrongShape.verdict).toBe(Verdict.Decline);
    expect(isReject(wrongShape.verdict)).toBe(true);

    // ── Site 3: fast-path selection (selectCertifiedBraidFastPath) ──────
    const certificate = createBraidFoldCertificate(exactCandidate);
    const chosen = selectCertifiedBraidFastPath(certificate);
    expect(chosen.verdict).toBe(Verdict.Accept);
    expect(chosen.selected).toBe(isAccept(chosen.verdict));

    // No certificate at all → nothing to select → Decline.
    expect(selectCertifiedBraidFastPath(null).verdict).toBe(Verdict.Decline);

    // Wankel fast path: absent → Decline, inexact → Decline.
    expect(selectWankelRuntimeFastPath(null).verdict).toBe(Verdict.Decline);
  });
});
