// @a0n/.ignored_gnosis/src/__tests__/study-graph-snapshot.test.ts
import { validateStudyGraphSnapshot } from '../study-graph-snapshot';

const valid = {
  schemaVersion: '1.0.0' as const,
  study_id: 'pilot-001',
  title: 'Qualitative pilot',
  exported_at: '2026-04-15T12:00:00.000Z',
  study_mode: 'confirmatory' as const,
  preregistration_ref: 'osf.io/abcd1',
  nodes: [
    { id: 'h1', kind: 'hypothesis' as const, label: 'H1' },
    { id: 'c1', kind: 'code' as const, label: 'theme-a' },
  ],
  edges: [{ u: 'h1', v: 'c1', provenance: 'human' as const }],
};

export default function law(): void {
  if (!validateStudyGraphSnapshot(valid).ok) {
    throw new Error('expected valid snapshot');
  }
  if (validateStudyGraphSnapshot({ ...valid, nodes: [{ id: 'x', kind: 'hypothesis' }, { id: 'x', kind: 'code' }] }).ok) {
    throw new Error('expected duplicate id rejection');
  }
  if (
    validateStudyGraphSnapshot({
      ...valid,
      edges: [{ u: 'h1', v: 'nope', provenance: 'human' as const }],
    }).ok
  ) {
    throw new Error('expected bad edge endpoint');
  }
}
