// @a0n/.ignored_gnosis/src/__tests__/study-publish-gate.test.ts
import { evaluateStudyPublishGate } from '../study-publish-gate';

const confirmatoryOk = {
  schemaVersion: '1.0.0' as const,
  study_id: 'pilot-001',
  title: 'Qualitative pilot',
  exported_at: '2026-04-15T12:00:00.000Z',
  study_mode: 'confirmatory' as const,
  preregistration_ref: 'osf.io/abcd1',
  nodes: [{ id: 'h1', kind: 'hypothesis' as const }],
  edges: [],
};

const exploratoryOk = {
  schemaVersion: '1.0.0' as const,
  study_id: 'exp-002',
  title: 'Exploratory memo',
  exported_at: '2026-04-15T12:00:00.000Z',
  study_mode: 'exploratory' as const,
  nodes: [{ id: 'm1', kind: 'memo' as const }],
  edges: [],
};

export default function law(): void {
  const r1 = evaluateStudyPublishGate({ studyGraphSnapshot: confirmatoryOk });
  if (!r1.ok) {
    throw new Error('expected confirmatory gate ok');
  }
  const r2 = evaluateStudyPublishGate({ studyGraphSnapshot: exploratoryOk });
  if (!r2.ok) {
    throw new Error('expected exploratory gate ok');
  }
  const bad = evaluateStudyPublishGate({
    studyGraphSnapshot: {
      ...confirmatoryOk,
      preregistration_ref: '',
    },
  });
  if (bad.ok) {
    throw new Error('expected prereg failure');
  }
  const noHyp = evaluateStudyPublishGate({
    studyGraphSnapshot: {
      ...confirmatoryOk,
      nodes: [{ id: 'c1', kind: 'code' as const }],
    },
  });
  if (noHyp.ok) {
    throw new Error('expected hypothesis requirement');
  }
}
