// @a0n/.ignored_gnosis/src/mod/loader.test.ts
import {
  mkdirSync,
  mkdtempSync,
  readFileSync,
  rmSync,
  writeFileSync,
} from 'node:fs';
import path from 'node:path';
import { pathToFileURL } from 'node:url';
import { afterEach, describe, expect, it } from '@a0n/gnosis/test';
import {
  GnosisModuleLoader,
  createFilesystemModuleResolver,
  loadGnosisModuleFromFile,
  parseMGG,
  renderMGG,
} from './loader';
import { ModManager } from './manager';

const tempDirs: string[] = [];
const tempRoot = path.join(process.cwd(), '.tmp-test-workspaces');

function makeTempDir(): string {
  mkdirSync(tempRoot, { recursive: true });
  const dir = mkdtempSync(path.join(tempRoot, 'gnosis-loader-'));
  tempDirs.push(dir);
  return dir;
}

afterEach(() => {
  while (tempDirs.length > 0) {
    const dir = tempDirs.pop();
    if (dir) {
      rmSync(dir, { recursive: true, force: true });
    }
  }
});

describe('parseMGG', () => {
  it('extracts imports, topology, and exports', () => {
    const parsed = parseMGG(`
import { router } from './routes.gg'

(request)-[:PROCESS]->(router)

export { request }
`);

    expect(parsed.imports).toEqual([
      {
        names: ['router'],
        source: './routes.gg',
      },
    ]);
    expect(parsed.topologySource).toContain('(request)-[:PROCESS]->(router)');
    expect(parsed.exports).toEqual([{ names: ['request'] }]);
  });

  it('renders a parsed module back to source', () => {
    const rendered = renderMGG({
      imports: [{ names: ['router'], source: './routes.gg' }],
      topologySource: '(request)-[:PROCESS]->(router)',
      exports: [{ names: ['request'] }],
    });

    expect(rendered).toContain("import { router } from './routes.gg'");
    expect(rendered).toContain('(request)-[:PROCESS]->(router)');
    expect(rendered).toContain('export { request }');
  });
});

describe('GnosisModuleLoader', () => {
  it('loads a module with a relative .gg import and merges its topology', async () => {
    const cwd = makeTempDir();
    const routesPath = path.join(cwd, 'routes.gg');
    const appPath = path.join(cwd, 'app.mgg');

    writeFileSync(
      routesPath,
      '(router:Router)\n(router)-[:PROCESS]->(handler:Handler)\n',
      'utf-8'
    );
    writeFileSync(
      appPath,
      [
        "import { router } from './routes.gg'",
        '',
        '(request:Request)',
        '(request)-[:PROCESS]->(router)',
        '',
        'export { request }',
        '',
      ].join('\n'),
      'utf-8'
    );

    const module = await loadGnosisModuleFromFile(appPath, cwd);

    expect(module.format).toBe('mgg');
    expect(module.exports).toEqual(['request']);
    expect(module.imports).toHaveLength(1);
    expect(module.ast.nodes.has('request')).toBe(true);
    expect(module.ast.nodes.has('router')).toBe(true);
    expect(
      [...module.ast.nodes.keys()].some((nodeId) => nodeId.endsWith(':handler'))
    ).toBe(true);
    expect(module.mergedSource).toContain('(request:Request)');
    expect(module.mergedSource).toContain('(request)-[:PROCESS]->(router)');
  });

  it('surfaces merged effect contracts for imported modules', async () => {
    const cwd = makeTempDir();
    const authPath = path.join(cwd, 'auth.gg');
    const appPath = path.join(cwd, 'app.mgg');

    writeFileSync(
      authPath,
      "(verify:UCANVerify { effects: 'auth.ucan' })\n",
      'utf-8'
    );
    writeFileSync(
      appPath,
      [
        "import { verify } from './auth.gg'",
        '',
        '(request:Request)',
        '(request)-[:PROCESS]->(verify)',
        '',
        'export { request }',
        '',
      ].join('\n'),
      'utf-8'
    );

    const module = await loadGnosisModuleFromFile(appPath, cwd);
    const verifyNode = module.effects.nodes.find(
      (node) => node.nodeId === 'verify'
    );

    expect(module.effects.required).toEqual(['auth.ucan']);
    expect(module.effects.declared).toEqual(['auth.ucan']);
    expect(module.effects.inferred).toEqual(['auth.ucan']);
    expect(verifyNode).toEqual({
      nodeId: 'verify',
      declared: ['auth.ucan'],
      inferred: ['auth.ucan'],
    });
  });

  it('supports empty re-export modules', async () => {
    const cwd = makeTempDir();
    const routesPath = path.join(cwd, 'routes.gg');
    const bridgePath = path.join(cwd, 'bridge.mgg');

    writeFileSync(
      routesPath,
      '(router:Router)\n(router)-[:PROCESS]->(handler:Handler)\n',
      'utf-8'
    );
    writeFileSync(
      bridgePath,
      [
        "import { router } from './routes.gg'",
        '',
        'export { router }',
        '',
      ].join('\n'),
      'utf-8'
    );

    const loader = new GnosisModuleLoader(createFilesystemModuleResolver(cwd));
    const module = await loader.load(bridgePath);

    expect(module.exports).toEqual(['router']);
    expect(module.ast.nodes.has('router')).toBe(true);
    expect(
      [...module.ast.nodes.keys()].some((nodeId) => nodeId.endsWith(':handler'))
    ).toBe(true);
  });

  it('loads plain gg modules through the unified format branch', async () => {
    const cwd = makeTempDir();
    const modulePath = path.join(cwd, 'plain.gg');

    writeFileSync(
      modulePath,
      [
        '(input:Tensor)',
        '(output:Tensor)',
        '(input)-[:PROCESS]->(output)',
        '',
      ].join('\n'),
      'utf-8'
    );

    const module = await loadGnosisModuleFromFile(modulePath, cwd);

    expect(module.format).toBe('gg');
    expect(module.imports).toEqual([]);
    expect(module.exports).toEqual(['input', 'output']);
    expect(module.ast.nodes.has('input')).toBe(true);
    expect(module.ast.nodes.has('output')).toBe(true);
    expect(module.mergedSource).toContain('(input)-[:PROCESS]->(output)');
  });

  it('keeps explicit export lists through the explicit-export branch', async () => {
    const cwd = makeTempDir();
    const modulePath = path.join(cwd, 'explicit-exports.mgg');

    writeFileSync(
      modulePath,
      [
        '(seed:Source)',
        '(seed)-[:PROCESS]->(sink:Sink)',
        '',
        'export { seed }',
        '',
      ].join('\n'),
      'utf-8'
    );

    const module = await loadGnosisModuleFromFile(modulePath, cwd);

    expect(module.exports).toEqual(['seed']);
    expect(module.ast.nodes.has('seed')).toBe(true);
    expect(module.ast.nodes.has('sink')).toBe(true);
  });

  it('loads standalone mgg modules through the no-import branch', async () => {
    const cwd = makeTempDir();
    const modulePath = path.join(cwd, 'standalone.mgg');

    writeFileSync(
      modulePath,
      [
        '(seed:Source)',
        '(seed)-[:PROCESS]->(sink:Sink)',
        '',
        'export { seed, sink }',
        '',
      ].join('\n'),
      'utf-8'
    );

    const module = await loadGnosisModuleFromFile(modulePath, cwd);

    expect(module.imports).toEqual([]);
    expect(module.exports).toEqual(['seed', 'sink']);
    expect(module.ast.nodes.has('seed')).toBe(true);
    expect(module.ast.nodes.has('sink')).toBe(true);
  });

  it('exports all local nodes when a module omits export declarations', async () => {
    const cwd = makeTempDir();
    const modulePath = path.join(cwd, 'implicit-exports.mgg');

    writeFileSync(
      modulePath,
      ['(seed:Source)', '(sink:Sink)', '(seed)-[:PROCESS]->(sink)', ''].join(
        '\n'
      ),
      'utf-8'
    );

    const module = await loadGnosisModuleFromFile(modulePath, cwd);

    expect(module.exports).toEqual(['seed', 'sink']);
    expect(module.ast.nodes.has('seed')).toBe(true);
    expect(module.ast.nodes.has('sink')).toBe(true);
  });

  it('supports imports from node-only .gg modules', async () => {
    const cwd = makeTempDir();
    const basePath = path.join(cwd, 'base.gg');
    const appPath = path.join(cwd, 'app.mgg');

    writeFileSync(
      basePath,
      "(input:Tensor { activation: 'identity' })\n",
      'utf-8'
    );
    writeFileSync(
      appPath,
      [
        "import { input } from './base.gg'",
        '',
        '(input)-[:PROCESS]->(output:Tensor)',
        '',
        'export { output }',
        '',
      ].join('\n'),
      'utf-8'
    );

    const module = await loadGnosisModuleFromFile(appPath, cwd);

    expect(module.ast.nodes.has('input')).toBe(true);
    expect(module.ast.nodes.has('output')).toBe(true);
    expect(module.mergedSource).toContain('(input:Tensor');
  });

  it('allows imports of target-only nodes from plain .gg files', async () => {
    const cwd = makeTempDir();
    const routesPath = path.join(cwd, 'routes.gg');
    const appPath = path.join(cwd, 'app.mgg');

    writeFileSync(
      routesPath,
      '(router:Router)\n(router)-[:PROCESS]->(step)-[:PROCESS]->(handler)\n',
      'utf-8'
    );
    writeFileSync(
      appPath,
      [
        "import { handler } from './routes.gg'",
        '',
        '(request:Request)',
        '(request)-[:PROCESS]->(handler)',
        '',
        'export { request }',
        '',
      ].join('\n'),
      'utf-8'
    );

    const module = await loadGnosisModuleFromFile(appPath, cwd);

    expect(module.ast.nodes.has('handler')).toBe(true);
    expect(module.mergedSource).toContain('(handler)');
  });

  it('rejects imports of non-exported symbols', async () => {
    const cwd = makeTempDir();
    const routesPath = path.join(cwd, 'routes.gg');
    const appPath = path.join(cwd, 'app.mgg');

    writeFileSync(
      routesPath,
      '(router:Router)\n(router)-[:PROCESS]->(handler:Handler)\n',
      'utf-8'
    );
    writeFileSync(
      appPath,
      [
        "import { missing } from './routes.gg'",
        '',
        '(request)-[:PROCESS]->(missing)',
        '',
      ].join('\n'),
      'utf-8'
    );

    await expect(loadGnosisModuleFromFile(appPath, cwd)).rejects.toThrow(
      "'missing' is not exported from './routes.gg'"
    );
  });

  it('resolves bare module specifiers through gnosis.lock dependencies', async () => {
    const cwd = makeTempDir();
    const appPath = path.join(cwd, 'app.mgg');
    const packageRoot = path.join(
      cwd,
      '.gnosis',
      'deps',
      'github.com',
      'acme',
      'routes',
      'v1.2.3'
    );

    writeFileSync(
      path.join(cwd, 'gnosis.mod'),
      [
        'module github.com/acme/app',
        '',
        'gnosis 0.1.0',
        '',
        'require github.com/acme/routes v1.2.3',
        '',
      ].join('\n'),
      'utf-8'
    );
    new ModManager(cwd).tidy();
    expect(readFileSync(path.join(cwd, 'gnosis.lock'), 'utf-8')).toContain(
      '"path": "github.com/acme/routes"'
    );

    mkdirSync(packageRoot, { recursive: true });
    writeFileSync(
      path.join(packageRoot, 'index.mgg'),
      [
        '(router:Router)',
        '(router)-[:PROCESS]->(handler:Handler)',
        '',
        'export { router }',
        '',
      ].join('\n'),
      'utf-8'
    );
    writeFileSync(
      appPath,
      [
        "import { router } from 'github.com/acme/routes'",
        '',
        '(request:Request)',
        '(request)-[:PROCESS]->(router)',
        '',
        'export { request }',
        '',
      ].join('\n'),
      'utf-8'
    );

    const module = await loadGnosisModuleFromFile(appPath, cwd);

    expect(module.imports).toHaveLength(1);
    expect(module.ast.nodes.has('router')).toBe(true);
    expect(
      [...module.ast.nodes.keys()].some((nodeId) => nodeId.endsWith(':handler'))
    ).toBe(true);
    expect(module.mergedSource).toContain('(request)-[:PROCESS]->(router)');
  });

  it('falls back to gnosis.mod dependencies when no lockfile exists yet', async () => {
    const cwd = makeTempDir();
    const appPath = path.join(cwd, 'app.mgg');
    const packageRoot = path.join(
      cwd,
      '.gnosis',
      'deps',
      'github.com',
      'acme',
      'routes',
      'v1.2.3'
    );

    writeFileSync(
      path.join(cwd, 'gnosis.mod'),
      [
        'module github.com/acme/app',
        '',
        'gnosis 0.1.0',
        '',
        'require github.com/acme/routes v1.2.3',
        '',
      ].join('\n'),
      'utf-8'
    );

    mkdirSync(packageRoot, { recursive: true });
    writeFileSync(
      path.join(packageRoot, 'index.gg'),
      '(router:Router)\n(router)-[:PROCESS]->(handler:Handler)\n',
      'utf-8'
    );
    writeFileSync(
      appPath,
      [
        "import { router } from 'github.com/acme/routes'",
        '',
        '(request)-[:PROCESS]->(router)',
        '',
        'export { request }',
        '',
      ].join('\n'),
      'utf-8'
    );

    const module = await loadGnosisModuleFromFile(appPath, cwd);

    expect(module.imports).toHaveLength(1);
    expect(module.ast.nodes.has('router')).toBe(true);
    expect(module.ast.nodes.has('request')).toBe(true);
  });

  it('rejects cyclic module imports with an import chain', async () => {
    const cwd = makeTempDir();
    const alphaPath = path.join(cwd, 'alpha.mgg');
    const betaPath = path.join(cwd, 'beta.mgg');

    writeFileSync(
      alphaPath,
      [
        "import { beta } from './beta.mgg'",
        '',
        '(alpha:Stage)',
        '(alpha)-[:PROCESS]->(beta)',
        '',
        'export { alpha }',
        '',
      ].join('\n'),
      'utf-8'
    );
    writeFileSync(
      betaPath,
      [
        "import { alpha } from './alpha.mgg'",
        '',
        '(beta:Stage)',
        '(beta)-[:PROCESS]->(alpha)',
        '',
        'export { beta }',
        '',
      ].join('\n'),
      'utf-8'
    );

    const loadResult = Promise.race([
      loadGnosisModuleFromFile(alphaPath, cwd),
      new Promise<never>((_, reject) => {
        setTimeout(() => {
          reject(new Error('module load timed out'));
        }, 1000);
      }),
    ]);

    await expect(loadResult).rejects.toThrow(
      /cyclic module import: .*alpha\.mgg -> .*beta\.mgg -> .*alpha\.mgg/
    );
  });
});

describe('createFilesystemModuleResolver', () => {
  it('resolves filesystem specifiers through the GG-native resolver topology', async () => {
    const cwd = makeTempDir();
    const routesPath = path.join(cwd, 'routes.gg');
    const appPath = path.join(cwd, 'app.mgg');

    writeFileSync(routesPath, '(router:Router)\n', 'utf-8');
    writeFileSync(appPath, '(request:Request)\n', 'utf-8');

    const resolver = createFilesystemModuleResolver(cwd);
    const result = await resolver('./routes', appPath);

    expect(result).toMatchObject({
      resolved: true,
      path: routesPath,
      source: '(router:Router)\n',
    });
  });

  it('resolves file URL specifiers through the GG-native resolver topology', async () => {
    const cwd = makeTempDir();
    const routesPath = path.join(cwd, 'routes.gg');

    writeFileSync(routesPath, '(router:Router)\n', 'utf-8');

    const resolver = createFilesystemModuleResolver(cwd);
    const result = await resolver(pathToFileURL(routesPath).href, '');

    expect(result).toMatchObject({
      resolved: true,
      path: routesPath,
      source: '(router:Router)\n',
    });
  });

  it('resolves bare specifiers through the GG-native resolver topology', async () => {
    const cwd = makeTempDir();
    const appPath = path.join(cwd, 'app.mgg');
    const dependencyRoot = path.join(
      cwd,
      '.gnosis',
      'deps',
      'github.com',
      'acme',
      'routes',
      'v1.2.3'
    );

    mkdirSync(dependencyRoot, { recursive: true });
    writeFileSync(
      path.join(dependencyRoot, 'index.gg'),
      '(router:Router)\n',
      'utf-8'
    );
    writeFileSync(appPath, '(request:Request)\n', 'utf-8');

    writeFileSync(
      path.join(cwd, 'gnosis.mod'),
      [
        'module github.com/acme/app',
        '',
        'gnosis 0.1.0',
        '',
        'require github.com/acme/routes v1.2.3',
        '',
      ].join('\n'),
      'utf-8'
    );
    new ModManager(cwd).tidy();

    const resolver = createFilesystemModuleResolver(cwd);
    const result = await resolver('github.com/acme/routes', appPath);

    expect(result).toMatchObject({
      resolved: true,
      path: path.join(dependencyRoot, 'index.gg'),
      source: '(router:Router)\n',
    });
  });

  it('falls back to the Forge registry for bare specifiers with channel aliases', async () => {
    const cwd = makeTempDir();
    const appPath = path.join(cwd, 'app.mgg');

    writeFileSync(appPath, '(request:Request)\n', 'utf-8');
    writeFileSync(
      path.join(cwd, 'gnosis.mod'),
      [
        'module github.com/acme/app',
        '',
        'gnosis 0.1.0',
        '',
        'require github.com/acme/routes latest',
        '',
      ].join('\n'),
      'utf-8'
    );
    new ModManager(cwd).tidy();

    const fetcher: typeof fetch = async (input) => {
      const url = new URL(String(input));
      if (
        url.pathname ===
        '/registry/gnosis/github.com%2Facme%2Froutes/channels/latest'
      ) {
        return new Response(
          JSON.stringify({
            moduleName: 'github.com/acme/routes',
            version: '0.1.0-forge.latest.20260327010101.deadbeefcafe',
            entryPath: 'index.mgg',
          }),
          {
            status: 200,
            headers: { 'content-type': 'application/json' },
          }
        );
      }

      if (
        url.pathname ===
        '/registry/gnosis/github.com%2Facme%2Froutes/versions/0.1.0-forge.latest.20260327010101.deadbeefcafe/files/index.mgg'
      ) {
        return new Response('(router:Router)\n', {
          status: 200,
          headers: { 'content-type': 'text/plain; charset=utf-8' },
        });
      }

      return new Response('not found', { status: 404 });
    };

    const resolver = createFilesystemModuleResolver(cwd, {
      registryBaseUrl: 'https://forgo.cloud/registry/gnosis',
      fetcher,
    });
    const result = await resolver('github.com/acme/routes', appPath);

    expect(result).toMatchObject({
      resolved: true,
      path: 'https://forgo.cloud/registry/gnosis/github.com%2Facme%2Froutes/versions/0.1.0-forge.latest.20260327010101.deadbeefcafe/files/index.mgg',
      source: '(router:Router)\n',
    });
  });
});
