// @a0n/gnosis/src/runtime/ambient-substrate.ts
import {
  gnosisAmbientRuntimeWorkIndex,
  gnosisAmbientRuntimeWorkOrder,
  planGnosisAmbientRuntimeSchedule,
  type GnosisAmbientRuntimeSchedule,
  type GnosisAmbientRuntimeScheduleInput,
} from './ambient-scheduler';

export const GNOSIS_AMBIENT_SUBSTRATES = Object.freeze([
  'polyglot',
  'gnode',
  'gnexec',
  'monster',
  'moonshine',
] as const);

export type GnosisAmbientSubstrate =
  (typeof GNOSIS_AMBIENT_SUBSTRATES)[number];

export interface GnosisAmbientSubstrateScheduleInput
  extends GnosisAmbientRuntimeScheduleInput {
  readonly substrate: GnosisAmbientSubstrate;
}

export interface GnosisAmbientSubstrateSchedule
  extends Omit<GnosisAmbientRuntimeSchedule, 'kind'> {
  readonly kind: 'gnosis-ambient-substrate-schedule';
  readonly substrate: GnosisAmbientSubstrate;
  readonly substrateRuntimeId: string;
  readonly admission: {
    readonly duplicatePayloads: false;
    readonly memoryFirst: true;
    readonly cryptographic: false;
  };
}

export interface GnosisAmbientSubstrateEnv {
  readonly GNOSIS_AMBIENT_SUBSTRATE: GnosisAmbientSubstrate;
  readonly GNOSIS_AMBIENT_RUNTIME_ID: string;
  readonly GNOSIS_AMBIENT_ENTROPY_WORD: string;
  readonly GNOSIS_AMBIENT_START_OFFSET: string;
  readonly GNOSIS_AMBIENT_STRIDE: string;
  readonly GNOSIS_AMBIENT_PARALLELISM: string;
  readonly GNOSIS_EDGEWORK_BOOST_UNITS: string;
  readonly GNOSIS_EDGEWORK_BOOST_LANES: string;
  readonly GNOSIS_AMBIENT_MAX_ITEMS_PER_CHUNK: string;
  readonly GNOSIS_AMBIENT_MEMORY_BUDGET_BYTES: string;
  readonly GNOSIS_AMBIENT_BYTES_PER_ITEM: string;
  readonly GNOSIS_AMBIENT_EXECUTION_PERMUTATION: string;
  readonly GNOSIS_MONSTER_RESIDENT_WINDOW: string;
}

/**
 * Handles the gnosis plan Gnosis Ambient Substrate Schedule workflow.
 */
export function planGnosisAmbientSubstrateSchedule(
  input: GnosisAmbientSubstrateScheduleInput
): GnosisAmbientSubstrateSchedule {
  assertAmbientSubstrate(input.substrate);
  const substrateRuntimeId =
    input.runtimeId?.trim() || `gnosis://${input.substrate}/ambient`;
  const runtime = planGnosisAmbientRuntimeSchedule({
    ...input,
    runtimeId: substrateRuntimeId,
    extraSources: [
      `substrate:${input.substrate}`,
      ...(input.extraSources ?? []),
    ],
  });

  return {
    ...runtime,
    kind: 'gnosis-ambient-substrate-schedule',
    substrate: input.substrate,
    substrateRuntimeId,
    admission: {
      duplicatePayloads: false,
      memoryFirst: true,
      cryptographic: false,
    },
  };
}

/**
 * Handles the gnosis gnosis Ambient Substrate Env workflow.
 */
export function gnosisAmbientSubstrateEnv(
  schedule: GnosisAmbientSubstrateSchedule
): GnosisAmbientSubstrateEnv {
  return {
    GNOSIS_AMBIENT_SUBSTRATE: schedule.substrate,
    GNOSIS_AMBIENT_RUNTIME_ID: schedule.substrateRuntimeId,
    GNOSIS_AMBIENT_ENTROPY_WORD: String(schedule.entropyWord),
    GNOSIS_AMBIENT_START_OFFSET: String(schedule.startOffset),
    GNOSIS_AMBIENT_STRIDE: String(schedule.stride),
    GNOSIS_AMBIENT_PARALLELISM: String(schedule.parallelism),
    GNOSIS_EDGEWORK_BOOST_UNITS: String(schedule.edgeworkBoostUnits),
    GNOSIS_EDGEWORK_BOOST_LANES: String(schedule.edgeworkBoostLanes),
    GNOSIS_AMBIENT_MAX_ITEMS_PER_CHUNK: String(schedule.maxItemsPerChunk),
    GNOSIS_AMBIENT_MEMORY_BUDGET_BYTES: String(schedule.memoryBudgetBytes),
    GNOSIS_AMBIENT_BYTES_PER_ITEM: String(schedule.bytesPerItem),
    GNOSIS_AMBIENT_EXECUTION_PERMUTATION:
      schedule.executionOrder.length === 0 ? '0' : '1',
    GNOSIS_MONSTER_RESIDENT_WINDOW: String(
      Math.max(1, Math.min(schedule.maxItemsPerChunk, schedule.parallelism))
    ),
  };
}

/**
 * Handles the gnosis gnosis Ambient Substrate Work Index workflow.
 */
export function gnosisAmbientSubstrateWorkIndex(
  schedule: Pick<GnosisAmbientSubstrateSchedule, 'startOffset' | 'stride'>,
  ordinal: number,
  workItems: number
): number {
  return gnosisAmbientRuntimeWorkIndex(schedule, ordinal, workItems);
}

/**
 * Handles the gnosis gnosis Ambient Substrate Work Order workflow.
 */
export function gnosisAmbientSubstrateWorkOrder(
  schedule: Pick<GnosisAmbientSubstrateSchedule, 'startOffset' | 'stride'>,
  workItems: number
): number[] {
  return [...gnosisAmbientRuntimeWorkOrder(schedule, workItems)];
}

function assertAmbientSubstrate(
  substrate: string
): asserts substrate is GnosisAmbientSubstrate {
  if (!GNOSIS_AMBIENT_SUBSTRATES.includes(substrate as GnosisAmbientSubstrate)) {
    throw new RangeError(`unknown Gnosis ambient substrate: ${substrate}`);
  }
}
