// @a0n/gnosis/src/runtime/amplituhedron-knot.ts
/**
 * Grassmannian Amplituhedron Knot Binary Format
 *
 * This codec replaces the legacy `packed-knot.ts` AST sequence.
 * Because we have collapsed the execution into a static volume,
 * we no longer serialize the Swarm's iterative steps.
 * 
 * Instead, the `.knot` file contains a single, fixed-size tensor:
 * The dimensions and scalar properties of the Amplituhedron.
 *
 * This achieves absolute $O(1)$ loading.
 *
 * Tensor Layout (20 bytes inner payload):
 *   0x00 - 0x03 : Origin Vertex (Clinamen) [Uint32]
 *   0x04 - 0x07 : Dimensionality (Bounds)  [Uint32]
 *   0x08 - 0x0F : Static Volume (Result)   [Float64]
 *   0x10 - 0x13 : Flattened Faces          [Uint32]
 */

import {
  MAGIC,
  writeHeader,
  readHeader,
  crc16,
  COMMON_HEADER_SIZE,
  verifyPayload,
} from '../lexicon';
import type { AmplituhedronGeometry } from '../betti/amplituhedron-volume';

export const AMPLITUHEDRON_PAYLOAD_SIZE = 20;

/**
 * Serialize the static Amplituhedron geometry into a pure binary `.knot` tensor.
 */
export function toBytes(geometry: AmplituhedronGeometry): Uint8Array {
  const innerPayload = new Uint8Array(AMPLITUHEDRON_PAYLOAD_SIZE);
  const view = new DataView(innerPayload.buffer);

  // Layout the tensor
  view.setUint32(0, geometry.originVertex, false);
  view.setUint32(4, geometry.dimensionality, false);
  view.setFloat64(8, geometry.staticVolume, false);
  view.setUint32(16, geometry.flattenedFaces, false);

  const crc = crc16(innerPayload);
  const buffer = new Uint8Array(COMMON_HEADER_SIZE + AMPLITUHEDRON_PAYLOAD_SIZE);

  // Write Lexicon Common Header
  writeHeader(buffer, MAGIC.GRSM, AMPLITUHEDRON_PAYLOAD_SIZE, {
    version: 1,
    flags: 0,
    crc,
  });

  buffer.set(innerPayload, COMMON_HEADER_SIZE);

  return buffer;
}

/**
 * Deserialize a pure binary `.knot` tensor into Amplituhedron geometry.
 */
export function fromBytes(bytes: Uint8Array): AmplituhedronGeometry {
  const header = readHeader(bytes);

  if (
    !header ||
    header.magic[0] !== MAGIC.GRSM[0] ||
    header.magic[1] !== MAGIC.GRSM[1]
  ) {
    throw new Error(`[Amplituhedron] Invalid knot magic: Expected GRSM tensor.`);
  }

  const innerPayload = bytes.subarray(
    COMMON_HEADER_SIZE,
    COMMON_HEADER_SIZE + header.payloadLength
  );

  if (!verifyPayload(header, innerPayload)) {
    throw new Error(`[Amplituhedron] CRC verification failed on geometry tensor.`);
  }

  const view = new DataView(
    innerPayload.buffer,
    innerPayload.byteOffset,
    innerPayload.byteLength
  );

  return {
    originVertex: view.getUint32(0, false),
    dimensionality: view.getUint32(4, false),
    staticVolume: view.getFloat64(8, false),
    flattenedFaces: view.getUint32(16, false),
  };
}

/**
 * Write a Grassmannian tensor to a VFS path as a .knot binary.
 */
export async function writebackToVFS(
  geometry: AmplituhedronGeometry,
  filePath: string
): Promise<void> {
  const { writeFile } = await import('node:fs/promises');
  await writeFile(filePath, toBytes(geometry));
}

/**
 * Load a Grassmannian tensor from a VFS path.
 */
export async function loadFromVFS(filePath: string): Promise<AmplituhedronGeometry> {
  const { readFile } = await import('node:fs/promises');
  const bytes = new Uint8Array(await readFile(filePath));
  return fromBytes(bytes);
}
