// @a0n/gnosis/src/runtime/core-cache.ts
import type { GraphAST } from '../betty/compiler';
import { QDoc } from '../crdt/index';
import {
  QCorridor,
  type QCorridorEntry,
  type QCorridorEvent,
  type QCorridorEvidence,
  type QCorridorExecutionOptions,
  type QCorridorLookup,
  type QCorridorMetrics,
  type QCorridorMode,
  type QCorridorObservation,
  type QCorridorRecord,
  type QCorridorReuseScope,
  type QCorridorSession,
} from '../crdt/qcorridor';

export type GnosisCoreCacheMode = QCorridorMode;
export type GnosisCoreCacheReuseScope = QCorridorReuseScope;
export type GnosisCoreCacheExecutionOptions = QCorridorExecutionOptions;
export type GnosisCoreCacheSession = QCorridorSession;
export type GnosisCoreCacheMetrics = QCorridorMetrics;
export type GnosisCoreCacheEntry = QCorridorEntry;
export type GnosisCoreCacheCorridorRecord = QCorridorRecord;
export type GnosisCoreCacheEvent = QCorridorEvent;
export type GnosisCoreCacheEvidence = QCorridorEvidence;
export type GnosisCoreCacheObservation = QCorridorObservation;
export type GnosisCoreCacheLookup<T> = QCorridorLookup<T>;

function stableStringify(value: unknown): string {
  if (value === null || value === undefined) {
    return 'null';
  }

  if (typeof value === 'string') {
    return JSON.stringify(value);
  }

  if (typeof value === 'number') {
    return Number.isFinite(value) ? String(value) : 'null';
  }

  if (typeof value === 'boolean') {
    return value ? 'true' : 'false';
  }

  if (typeof value === 'bigint') {
    return JSON.stringify(value.toString(10));
  }

  if (Array.isArray(value)) {
    return `[${value.map((item) => stableStringify(item)).join(',')}]`;
  }

  if (typeof value === 'object') {
    const record = value as Record<string, unknown>;
    const entries = Object.entries(record)
      .filter(([, entryValue]) => {
        const entryType = typeof entryValue;
        return (
          entryType !== 'undefined' &&
          entryType !== 'function' &&
          entryType !== 'symbol'
        );
      })
      .sort(([leftKey], [rightKey]) => leftKey.localeCompare(rightKey));

    return `{${entries
      .map(
        ([key, entryValue]) =>
          `${JSON.stringify(key)}:${stableStringify(entryValue)}`
      )
      .join(',')}}`;
  }

  return 'null';
}

function serializeAst(ast: GraphAST): string {
  const nodes = [...ast.nodes.values()]
    .map((node) => ({
      id: node.id,
      labels: [...node.labels],
      properties: node.properties,
    }))
    .sort((left, right) => left.id.localeCompare(right.id));

  const edges = [...ast.edges]
    .map((edge) => ({
      sourceIds: [...edge.sourceIds],
      targetIds: [...edge.targetIds],
      type: edge.type,
      properties: edge.properties,
    }))
    .sort((left, right) => {
      const leftKey = `${left.type}:${left.sourceIds.join(
        '|'
      )}->${left.targetIds.join('|')}:${stableStringify(left.properties)}`;
      const rightKey = `${right.type}:${right.sourceIds.join(
        '|'
      )}->${right.targetIds.join('|')}:${stableStringify(right.properties)}`;
      return leftKey.localeCompare(rightKey);
    });

  return stableStringify({ nodes, edges });
}

export class GnosisCoreCache {
  private readonly corridor: QCorridor;

  constructor(doc: QDoc = new QDoc({ guid: 'gnosis-core-cache' })) {
    this.corridor = new QCorridor(doc, { namespace: 'runtime' });
  }

  get document(): QDoc {
    return this.corridor.document;
  }

  getCorridor(corridorKey: string): GnosisCoreCacheCorridorRecord | undefined {
    return this.corridor.getCorridor(corridorKey);
  }

  getEntry(requestKey: string): GnosisCoreCacheEntry | undefined {
    return this.corridor.getEntry(requestKey);
  }

  getEvents(): GnosisCoreCacheEvent[] {
    return this.corridor.getEvents();
  }

  getEvidence(): GnosisCoreCacheEvidence[] {
    return this.corridor.getEvidence();
  }

  clearInflight(): void {
    this.corridor.clearInflight();
  }

  async createSession(
    ast: GraphAST,
    payload: unknown,
    options: GnosisCoreCacheExecutionOptions = {}
  ): Promise<GnosisCoreCacheSession> {
    return this.corridor.createSession(serializeAst(ast), payload, options);
  }

  lookup<T>(session: GnosisCoreCacheSession): GnosisCoreCacheLookup<T> {
    return this.corridor.lookup<T>(session);
  }

  registerInflight<T>(
    session: GnosisCoreCacheSession,
    promise: Promise<T>
  ): void {
    this.corridor.registerInflight(session, promise);
  }

  release(session: GnosisCoreCacheSession): void {
    this.corridor.release(session);
  }

  commit(
    session: GnosisCoreCacheSession,
    payload: unknown,
    metrics: GnosisCoreCacheMetrics
  ): GnosisCoreCacheEntry {
    return this.corridor.settle(session, payload, metrics);
  }

  fail(session: GnosisCoreCacheSession, error: unknown): void {
    this.corridor.fail(session, error);
  }

  recordEvidence(
    session: GnosisCoreCacheSession,
    observations: readonly GnosisCoreCacheObservation[]
  ): GnosisCoreCacheEvidence[] {
    return this.corridor.recordEvidence(session, observations);
  }

  readPayload(entry: GnosisCoreCacheEntry): {
    found: boolean;
    payload: unknown;
  } {
    return this.corridor.readPayload(entry);
  }

  corridorStrength(corridorKey: string): number {
    return this.corridor.getCorridorStrength(corridorKey);
  }

  rankPaths(
    session: GnosisCoreCacheSession,
    paths: readonly string[]
  ): string[] {
    return this.corridor.rankPaths(session.corridorKey, paths);
  }
}
