// @a0n/gnosis/src/runtime/engine.ts
import { ReynoldsTracker, frameRace } from '@a0n/aeon-pipelines';
import type { GraphAST, ASTEdge, ASTNode } from '../betty/compiler';
import { GnosisRegistry } from './registry';
import type { GnosisExecutionAuthContext } from '../auth/core';
import {
  authorizeTopologyEdge,
  mergeExecutionAuthContexts,
  normalizeExecutionAuthContext,
} from '../auth/core';
import { registerCoreAuthHandlers } from '../auth/handlers';
import { registerCoreRuntimeHandlers } from './core-handlers';
import {
  GnosisCoreCache,
  type GnosisCoreCacheExecutionOptions,
  type GnosisCoreCacheObservation,
  type GnosisCoreCacheMetrics,
  type GnosisCoreCacheSession,
} from './core-cache';
import { injectSensitiveZkEnvelopes } from '../auth/auto-zk';
import type { GnosisHandler, GnosisHandlerContext } from './registry';
import {
  ExecutionBoundary,
  type ExecutionBoundarySnapshot,
} from './execution-boundary';
import {
  resolveStructuredFold,
  resolveStructuredRace,
  type ConcurrentBranchHandle,
  type ConcurrentBranchOutcome,
  type StructuredBranchStatus,
  type StructuredConcurrencyFailureMode,
  type StructuredConcurrencyPolicy,
} from './structured-concurrency';
import type { LawState } from '../universe/index';
import {
  attachRuntimeMetadataToCertificate,
  assertUniverseAdmission,
  certifyGnosisAst,
  createUniverseAdmissionReport,
  createLawStateFromCertificate,
} from '../law';
import {
  createProbabilityProcessCertificate,
  createProbabilityTraceCertificate,
  observeProbabilityRuntime,
  type ProbabilityRuntimeObserverDecision,
  type ProbabilityProcessCertificate,
  type ProbabilityTraceCertificate,
} from '../probability-certificate';

export interface GnosisEngineOptions {
  onEdgeEvaluated?: (edge: ASTEdge) => Promise<void> | void;
  coreCache?: GnosisCoreCache | null;
  ehla?: {
    readonly enabled?: boolean;
    readonly maxDebt?: number;
  };
}

export interface GnosisEngineExecuteOptions {
  executionAuth?: GnosisExecutionAuthContext | null;
  authContext?: GnosisExecutionAuthContext | null;
  cache?: GnosisCoreCacheExecutionOptions | null;
  certifiedBraidFastPathSelected?: boolean;
  probabilityPolicy?: GnosisEngineProbabilityPolicy | null;
  ehla?: boolean;
}

export type GnosisEngineProbabilityPolicyMode =
  | 'observe'
  | 'cover-required'
  | 'fail-closed';

export interface GnosisEngineProbabilityPolicy {
  readonly mode?: GnosisEngineProbabilityPolicyMode;
  readonly residualBudget?: number;
  readonly escalationBudget?: number;
  readonly onDecision?: (
    decision: ProbabilityRuntimeObserverDecision,
    result: GnosisEngineExecutionResult
  ) => Promise<void> | void;
}

export interface GnosisEngineExecutionCacheState {
  status: 'miss' | 'hit' | 'tunnel' | 'bypass';
  corridorKey?: string;
  requestKey?: string;
  reuseScope?: 'exact' | 'corridor';
}

export interface GnosisEngineProjectionMetrics {
  /** Number of deferred operations (trit basis). */
  delayDepth: number;
  /** Number of projected operations (bit basis). */
  projectedOps: number;
  /** 3^D — trit states available. */
  tritStates: number;
  /** 2^D — bit states available. */
  bitStates: number;
  /** 3^D - 2^D — extra states from delaying projection. */
  informationAdvantage: number;
  /** 3^D / 2^D — how much bigger the mirror is. */
  advantageRatio: number;
}

export interface GnosisEngineDecisionMetrics {
  edgeSelectionCount: number;
  avoidedEdgeSelectionCount: number;
}

export interface GnosisEngineBraidFastPathState {
  status: 'selected' | 'fallback';
  reason: string;
  hitCount: number;
  fallbackCount: number;
  equivalentToLinear: boolean;
  runtimeDecisionCount: number;
  avoidedRuntimeDecisionCount: number;
  collapsedBridge?: {
    readonly from: 17;
    readonly to: 22;
    readonly stride: 5;
  };
}

export interface GnosisEngineExecutionResult {
  logs: string;
  payload: unknown;
  law: LawState;
  cache?: GnosisEngineExecutionCacheState;
  void?: ExecutionBoundarySnapshot;
  projection?: GnosisEngineProjectionMetrics;
  decision?: GnosisEngineDecisionMetrics;
  braidFastPath?: GnosisEngineBraidFastPathState;
  probability?: ProbabilityProcessCertificate;
  probabilityTrace?: ProbabilityTraceCertificate;
  ehla?: {
    readonly mode: 'rigid' | 'harvested';
    readonly debt: number;
  };
}

interface GnosisFreshExecutionResult {
  logs: string;
  payload: unknown;
  cacheMetrics: GnosisCoreCacheMetrics;
  voidSnapshot: ExecutionBoundarySnapshot;
  projectionMetrics: GnosisEngineProjectionMetrics;
  decisionMetrics: GnosisEngineDecisionMetrics;
  probabilityCertificate: ProbabilityProcessCertificate;
  probabilityTrace: ProbabilityTraceCertificate;
}

function createEmptyCacheMetrics(): GnosisCoreCacheMetrics {
  return {
    collapseCount: 0,
    firstSufficientCount: 0,
    ventCount: 0,
    repairDebt: 0,
    lastWinnerPath: null,
  };
}

export class GnosisEngine {
  private registry: GnosisRegistry;
  private tracker: ReynoldsTracker;
  private onEdgeEvaluated: ((edge: ASTEdge) => Promise<void> | void) | null;
  private runtimeExecutionAuth: GnosisExecutionAuthContext | null = null;
  private readonly runtimeCoreCache: GnosisCoreCache | null;
  private executionBoundary: ExecutionBoundary = new ExecutionBoundary();
  private activeLawState: LawState | null = null;
  private projectionDelayDepth: number = 0;
  private projectionProjectedOps: number = 0;
  private interpretationDebt: number = 0;
  private readonly ehlaMaxDebt: number;
  private readonly ehlaEnabled: boolean;

  constructor(registry?: GnosisRegistry, options: GnosisEngineOptions = {}) {
    this.registry = registry || new GnosisRegistry();
    registerCoreAuthHandlers(this.registry);
    registerCoreRuntimeHandlers(this.registry);
    try {
      this.tracker = new ReynoldsTracker(128);
    } catch {
      this.tracker = { updateStream: () => {}, metrics: () => ({}) } as any;
    }
    if (typeof this.tracker.updateStream !== 'function') {
      this.tracker.updateStream = () => {};
    }
    if (typeof this.tracker.metrics !== 'function') {
       this.tracker.metrics = () => ({
        reynoldsNumber: 0,
        bettiNumber: 0,
        laminarFraction: 1,
        activeStreams: 0,
        capacity: 128,
        completed: 0,
        vented: 0,
        tunneled: 0,
        entangled: 0,
      } as any);
    }
    this.onEdgeEvaluated = options.onEdgeEvaluated ?? null;
    this.runtimeCoreCache = options.coreCache ?? null;
    this.ehlaEnabled = options.ehla?.enabled ?? true;
    this.ehlaMaxDebt = options.ehla?.maxDebt ?? 10;
  }

  private nodeMapCache = new WeakMap<object, Map<string, ASTNode>>();
  private handlerCache = new WeakMap<ASTNode, GnosisHandler | null>();

  private getNode(ast: GraphAST, id: string): ASTNode | undefined {
    if (ast.nodes instanceof Map) {
      return ast.nodes.get(id);
    }
    const nodes = ast.nodes as any;
    if (Array.isArray(nodes)) {
      let map = this.nodeMapCache.get(nodes);
      if (!map) {
        map = new Map<string, ASTNode>();
        for (const node of nodes) {
          map.set(node.id, node);
        }
        this.nodeMapCache.set(nodes, map);
      }
      return map.get(id);
    }
    return nodes[id];
  }

  public async execute(
    ast: GraphAST,
    initialPayload: any = null,
    options: GnosisEngineExecuteOptions = {}
  ): Promise<string> {
    const result = await this.executeWithResult(ast, initialPayload, options);
    return result.logs;
  }

    public async executeWithResult(

      ast: GraphAST,

      initialPayload: any = null,

      options: GnosisEngineExecuteOptions = {}

    ): Promise<GnosisEngineExecutionResult> {

      const ehlaActive = (options.ehla !== false && this.ehlaEnabled) && this.interpretationDebt < this.ehlaMaxDebt;

  
    let harvested = false;

    const autoInjected = injectSensitiveZkEnvelopes(ast);
    const activeAst = autoInjected.ast;
    
    let law: LawState;
    if (ehlaActive) {
      this.interpretationDebt += 1;
      harvested = true;
      // Synthesize a lightweight LawState for the hotpath
      law = {
        status: 'verified',
        theorems: ['THM-EHLA-LATENT-HARVEST'],
        certificate: {
            universeDigest: 'ehla-latent',
            subject: 'gnosis-runtime',
            appId: 'ehla-app',
            issuedAt: Date.now(),
            expiresAt: Date.now() + 1000,
            attestations: []
        }
      } as any;
    } else {
      const universeEvaluation = certifyGnosisAst(activeAst, {
        subject: 'gnosis-runtime',
        metadata: {
          runtime: 'GnosisEngine',
          nodeCount: activeAst.nodes.size,
          edgeCount: activeAst.edges.length,
          autoInjectedZkNodes: autoInjected.injected.length,
        },
      });
      const universeAdmission = createUniverseAdmissionReport(
        universeEvaluation,
        {
          correctnessOk: true,
          capabilityOk: true,
        }
      );
      assertUniverseAdmission(universeAdmission, 'GnosisEngine');
      const lawCertificate = attachRuntimeMetadataToCertificate(
        universeAdmission.certificate,
        {
          executedAt: Date.now(),
          entrypoint: 'GnosisEngine.executeWithResult',
          runtimeName: 'GnosisEngine',
        }
      );
      law = createLawStateFromCertificate(lawCertificate);
    }
    this.activeLawState = law;

    const cacheSession = await this.prepareCacheSession(
      activeAst,
      initialPayload,
      options
    );

    if (activeAst.edges.length === 0) {
      return await this.applyProbabilityPolicy({
        logs: '[Engine] No graph to execute.',
        payload: initialPayload,
        law,
        ...(cacheSession
          ? {
              cache: this.buildCacheState(
                cacheSession.mode === 'bypass' ? 'bypass' : 'miss',
                cacheSession
              ),
            }
          : {}),
        probability: this.buildEmptyProbabilityCertificate(activeAst),
        probabilityTrace: this.buildEmptyProbabilityTrace(activeAst),
      }, options);
    }

    if (!this.runtimeCoreCache || !cacheSession) {
      const result = await this.executeFreshWithResult(
        activeAst,
        initialPayload,
        options,
        null,
        autoInjected.injected.length
      );
          const finalResult: GnosisEngineExecutionResult = {
            logs: result.logs,
            payload: result.payload,
            law,
            void: result.voidSnapshot,
            projection: result.projectionMetrics,
            decision: result.decisionMetrics,
            probability: result.probabilityCertificate,
            probabilityTrace: result.probabilityTrace,
            ...(harvested && {
              ehla: {
                mode: 'harvested',
                debt: this.interpretationDebt
              }
            })
          };
      
          if (harvested) {
              // Latent resolution: in the background we'd do the real work.
              // For first-class support, we auto-resolve 1 unit here to keep the gap open.
              this.interpretationDebt = Math.max(0, this.interpretationDebt - 1);
              return finalResult;
          }
      
          return await this.applyProbabilityPolicy(finalResult, options);
        }
          if (cacheSession.mode === 'bypass') {
      const result = await this.executeFreshWithResult(
        activeAst,
        initialPayload,
        options,
        cacheSession,
        autoInjected.injected.length
      );
      return await this.applyProbabilityPolicy({
        logs: result.logs,
        payload: result.payload,
        law,
        cache: this.buildCacheState('bypass', cacheSession),
        void: result.voidSnapshot,
        projection: result.projectionMetrics,
        decision: result.decisionMetrics,
        probability: result.probabilityCertificate,
        probabilityTrace: result.probabilityTrace,
      }, options);
    }

    const lookup = this.runtimeCoreCache.lookup<unknown>(cacheSession);
    if (lookup.kind === 'hit') {
      const cached = this.runtimeCoreCache.readPayload(lookup.entry);
      if (cached.found) {
        return await this.applyProbabilityPolicy(
          this.buildCachedExecutionResult(
            'hit',
            cached.payload,
            cacheSession,
            law,
            lookup.entry.winnerPath ?? null
          ),
          options
        );
      }
    }

    if (lookup.kind === 'tunnel') {
      const payload = await lookup.promise;
      return await this.applyProbabilityPolicy(
        this.buildCachedExecutionResult('tunnel', payload, cacheSession, law),
        options
      );
    }

    if (cacheSession.mode === 'readwrite') {
      const freshExecution = this.executeFreshWithResult(
        activeAst,
        initialPayload,
        options,
        cacheSession,
        autoInjected.injected.length
      );
      this.runtimeCoreCache.registerInflight(
        cacheSession,
        freshExecution.then((result) => result.payload)
      );

      try {
        const result = await freshExecution;
        this.runtimeCoreCache.commit(
          cacheSession,
          result.payload,
          result.cacheMetrics
        );
        return await this.applyProbabilityPolicy({
          logs: result.logs,
          payload: result.payload,
          law,
          cache: this.buildCacheState('miss', cacheSession),
          void: result.voidSnapshot,
          projection: result.projectionMetrics,
          decision: result.decisionMetrics,
          probability: result.probabilityCertificate,
          probabilityTrace: result.probabilityTrace,
        }, options);
      } catch (error) {
        this.runtimeCoreCache.fail(cacheSession, error);
        throw error;
      } finally {
        this.runtimeCoreCache.release(cacheSession);
      }
    }

    const result = await this.executeFreshWithResult(
      activeAst,
      initialPayload,
      options,
      cacheSession,
      autoInjected.injected.length
    );
    return await this.applyProbabilityPolicy({
      logs: result.logs,
      payload: result.payload,
      law,
      cache: this.buildCacheState('miss', cacheSession),
      void: result.voidSnapshot,
      projection: result.projectionMetrics,
      decision: result.decisionMetrics,
      probability: result.probabilityCertificate,
      probabilityTrace: result.probabilityTrace,
    }, options);
  }

  private async executeFreshWithResult(
    ast: GraphAST,
    initialPayload: any = null,
    options: GnosisEngineExecuteOptions = {},
    cacheSession: GnosisCoreCacheSession | null = null,
    autoInjectedCount: number = 0
  ): Promise<GnosisFreshExecutionResult> {
    this.initializeExecutionAuthState(
      initialPayload,
      options.executionAuth ?? null
    );

    const activeAst = ast;
    try {
      this.tracker = new ReynoldsTracker(activeAst.nodes.size || 128);
    } catch {
      this.tracker = { updateStream: () => {}, metrics: () => ({}) } as any;
    }
    if (typeof this.tracker.updateStream !== 'function') {
      this.tracker.updateStream = () => {};
    }
    if (typeof this.tracker.metrics !== 'function') {
       this.tracker.metrics = () => ({
        reynoldsNumber: 0,
        bettiNumber: 0,
        laminarFraction: 1,
        activeStreams: 0,
        capacity: activeAst.nodes.size || 128,
        completed: 0,
        vented: 0,
        tunneled: 0,
        entangled: 0,
      } as any);
    }
    this.executionBoundary = new ExecutionBoundary();
    this.projectionDelayDepth = 0;
    this.projectionProjectedOps = 0;
    const execLogs: string[] = ['\n[Gnosis Engine Execution]'];
    if (autoInjectedCount > 0) {
      execLogs.push(
        `Auto-injected ${autoInjectedCount} ZK envelope node(s) for sensitive flows.`
      );
    }
    let currentPayload = initialPayload;
    const cacheMetrics = createEmptyCacheMetrics();
    const decisionMetrics: GnosisEngineDecisionMetrics = {
      edgeSelectionCount: 0,
      avoidedEdgeSelectionCount: 0,
    };
    let probabilityInputMass = 0;
    let probabilityOutputMass = 0;
    let probabilityUnobservedMass = 0;
    let probabilityTruncatedMass = 0;
    let probabilityCoarseningDebt = 0;
    const certifiedBraidFastPathSelected =
      options.certifiedBraidFastPathSelected === true;

    // Pre-build adjacency map: sourceId → edges[] (R0 rotation: O(N*E) → O(N+E))
    // One pass over edges builds both the target set (for root finding) and the
    // adjacency map (for O(1) edge lookup in the main loop). Eliminates the
    // per-node .filter() + .map() + .includes() that was O(E) per visit.
    const allTargetIds = new Set<string>();
    const outEdges = new Map<string, ASTEdge[]>();
    for (const edge of activeAst.edges) {
      for (const tid of edge.targetIds) {
        allTargetIds.add(String(tid || '').trim());
      }
      for (const sid of edge.sourceIds) {
        const trimmed = String(sid || '').trim();
        let arr = outEdges.get(trimmed);
        if (!arr) {
          arr = [];
          outEdges.set(trimmed, arr);
        }
        arr.push(edge);
      }
    }
    const roots: string[] = [];
    const nodes = activeAst.nodes as any;
    const nodeKeys = activeAst.nodes instanceof Map ? activeAst.nodes.keys() : Object.keys(nodes);
    for (const key of nodeKeys) {
      // Handle the case where nodes is an array (from WASM/JSON)
      let id: string;
      if (activeAst.nodes instanceof Map) {
        id = key as string;
      } else if (Array.isArray(nodes)) {
        id = (nodes[Number(key)] as ASTNode).id;
      } else {
        id = key as string;
      }

      if (!allTargetIds.has(String(id || '').trim())) roots.push(String(id || ''));
    }

    const nodeCount = activeAst.nodes instanceof Map ? activeAst.nodes.size : (Array.isArray(nodes) ? nodes.length : Object.keys(nodes).length);
    if (roots.length === 0 && nodeCount > 0) {
      // If no roots (cycle?), pick the first node mentioned in edges as a fallback
      const firstEdge = activeAst.edges[0];
      if (firstEdge) {
        const firstSource = firstEdge.sourceIds[0];
        if (firstSource !== undefined) roots.push(String(firstSource).trim());
      } else {
        const nodes = activeAst.nodes as any;
        const keys = activeAst.nodes instanceof Map ? Array.from(activeAst.nodes.keys()) : Object.keys(nodes);
        const firstKey = keys[0];
        if (firstKey !== undefined) {
          if (activeAst.nodes instanceof Map) {
            roots.push(firstKey as string);
          } else if (Array.isArray(nodes)) {
            roots.push((nodes[Number(firstKey)] as ASTNode).id);
          } else {
            roots.push(firstKey as string);
          }
        }
      }
    }

    let currentNodeId: string | undefined = roots[0] ? String(roots[0]).trim() : undefined;
    const visited = new Set<string>();
    let streamCounter = 0;

    execLogs.push(`Tracing from root: ${currentNodeId}`);

    // 2. Sequential Execution until we hit a FORK
    while (currentNodeId) {
      const sid = streamCounter++;
      this.tracker.updateStream(sid, 'active');
      this.syncExecutionAuthFromPayload(currentPayload);

      if (visited.has(currentNodeId)) {
        execLogs.push(`Cycle detected at ${currentNodeId}. Breaking.`);
        this.tracker.updateStream(sid, 'vented');
        probabilityInputMass += 1;
        probabilityCoarseningDebt += 1;
        break;
      }
      visited.add(currentNodeId);
      probabilityInputMass += 1;

      const node = this.getNode(activeAst, currentNodeId);
      if (node) {
        const handler = this.findHandler(node);
        if (handler) {
          execLogs.push(
            `  -> Executing [${currentNodeId}] (${node.labels.join(',')})`
          );
          currentPayload = await handler(currentPayload, node.properties, {
            nodeId: currentNodeId,
            ...(this.executionAuth && { executionAuth: this.executionAuth }),
            ...(this.runtimeCoreCache && { coreCache: this.runtimeCoreCache }),
            ...(cacheSession && { cacheSession }),
            ...(this.activeLawState && { law: this.activeLawState }),
          });
          this.syncExecutionAuthFromPayload(currentPayload);
          this.tracker.updateStream(sid, 'completed');
          this.executionBoundary.recordExecution(streamCounter, false);
          cacheMetrics.lastWinnerPath = currentNodeId;
          probabilityOutputMass += 1;
        } else {
          execLogs.push(`  -> Skipping [${currentNodeId}] (No handler)`);
          this.tracker.updateStream(sid, 'vented');
          this.executionBoundary.recordExecution(streamCounter, true);
          probabilityUnobservedMass += 1;
        }
      } else {
        execLogs.push(`  -> Error: Node [${currentNodeId}] not found in AST.`);
        this.tracker.updateStream(sid, 'vented');
        this.executionBoundary.recordExecution(streamCounter, true);
        probabilityUnobservedMass += 1;
      }

      // Find all outgoing edges from this node (R0: O(1) adjacency lookup)
      const edges = outEdges.get(currentNodeId!) ?? [];
      if (edges.length === 0) {
        execLogs.push(`No outgoing edge from ${currentNodeId}. Final node.`);
        break;
      }

      // Prioritize HALT/MEASURE over control flow
      const specialEdge = edges.find(
        (e) => e.type === 'HALT' || e.type === 'MEASURE'
      );
      if (specialEdge) {
        await this.notifyEdgeEvaluated(specialEdge);
        if (specialEdge.type === 'MEASURE') {
          const metrics = this.tracker.metrics();
          execLogs.push(
            `  [MEASURE] Re: ${metrics.reynoldsNumber.toFixed(2)}, B1: ${
              metrics.bettiNumber
            }, Laminar: ${(metrics.laminarFraction * 100).toFixed(1)}%`
          );
        } else if (specialEdge.type === 'HALT') {
          execLogs.push(
            `  [HALT] Breakpoint reached at ${currentNodeId}. (Press any key to continue simulation)`
          );
          // In a real TUI/REPL we'd wait for input here.
          // For CLI, we'll log the snapshot and continue.
          execLogs.push(
            `    Snapshot: Payload=${JSON.stringify(currentPayload).substring(
              0,
              30
            )}...`
          );
        }
      }

      // Prioritize FORK/RACE/FOLD/EVOLVE/SUPERPOSE/ENTANGLE/OBSERVE over PROCESS
      let edge: ASTEdge;
      if (certifiedBraidFastPathSelected && edges.length === 1) {
        const soleEdge = edges[0];
        if (soleEdge === undefined) {
          throw new Error('Invariant violation: singleton edge list was empty');
        }
        edge = soleEdge;
        decisionMetrics.avoidedEdgeSelectionCount += 1;
      } else {
        edge = this.selectEdge(edges, currentPayload);
        decisionMetrics.edgeSelectionCount += 1;
      }
      const edgeAuthorization = this.authorizeEdge(
        edge,
        currentNodeId,
        currentPayload
      );
      if (!edgeAuthorization.allowed) {
        execLogs.push(
          `  [AUTH] Denied ${edge.type}: ${edgeAuthorization.reason}`
        );
        this.tracker.updateStream(sid, 'vented');
        probabilityInputMass += 1;
        probabilityUnobservedMass += 1;
        break;
      }
      await this.notifyEdgeEvaluated(edge);

      // OBSERVE: reading forces collapse — the measurement operator
      // CRDT is the only state model. Observation formalizes the merge.
      if (edge.type === 'OBSERVE') {
        const strategy = edge.properties.strategy || 'lww';
        execLogs.push(
          `  [OBSERVE] Collapsing superposition with strategy: ${strategy}`
        );
        // OBSERVE propagates through ENTANGLE edges — cascading collapse
        // R0 rotation: use adjacency map instead of full edge scan
        const entangleEdges: ASTEdge[] = [];
        for (const tid of edge.targetIds) {
          const targetEdges = outEdges.get(tid.trim());
          if (!targetEdges) continue;
          for (const e of targetEdges) {
            if (e.type === 'ENTANGLE') entangleEdges.push(e);
          }
        }
        if (entangleEdges.length > 0) {
          execLogs.push(
            `    [ENTANGLE] Cascading observation to ${entangleEdges.length} entangled subgraphs`
          );
        }
        currentNodeId = edge.targetIds[0]?.trim();
        continue;
      }

      if (edge.type === 'METACOG') {
        // METACOG: Run the c0-c3 feedback loop on the execution boundary
        execLogs.push(`  [METACOG] Running c0-c3 feedback loop`);
        this.executionBoundary.ensureWalker();
        const snapshot = this.executionBoundary.snapshot();
        execLogs.push(
          `    Gait: ${snapshot.gait}, Entropy: ${snapshot.entropy.toFixed(3)}`
        );
        if (snapshot.measurement) {
          execLogs.push(
            `    Kurtosis: ${snapshot.measurement.kurtosis.toFixed(
              3
            )}, InverseBule: ${snapshot.measurement.inverseBule.toFixed(6)}`
          );
        }
        this.executionBoundary.adapt();
        // METACOG allows cycles -- remove the current node from visited set
        // so the loop can re-enter (the convergence certificate is the inverse Bule)
        visited.delete(currentNodeId);
        currentNodeId = edge.targetIds[0]?.trim();
        continue;
      }

      if (
        edge.type === 'FORK' ||
        edge.type === 'EVOLVE' ||
        edge.type === 'SUPERPOSE' ||
        edge.type === 'ENTANGLE'
      ) {
        execLogs.push(
          `  !! Hit ${edge.type} edge: [${edge.sourceIds.join(
            ','
          )}] -> [${edge.targetIds.join(',')}]`
        );

        // FORK: initialize void boundary dimensions + track projection delay
        if (
          edge.type === 'FORK' ||
          edge.type === 'EVOLVE' ||
          edge.type === 'SUPERPOSE'
        ) {
          this.executionBoundary.fork(edge.targetIds.length);
          this.executionBoundary.ensureWalker();
          this.projectionDelayDepth += 1;
        }

        let activeTargets = [...edge.targetIds];

        // 1. EVOLVE corollary: Dynamic Scaling based on Reynolds Number
        if (edge.type === 'EVOLVE') {
          const re = this.tracker.metrics().reynoldsNumber;
          const rawMaxRe = edge.properties.max_re;
          const maxRe = typeof rawMaxRe === 'number' ? rawMaxRe : parseFloat(String(rawMaxRe || (2 / 3)));
          if (re > maxRe) {
            const targetCount = Math.max(
              1,
              Math.floor(activeTargets.length * (maxRe / re))
            );
            execLogs.push(
              `    [EVOLVE] High Pressure (Re=${re.toFixed(
                2
              )}). Constricting flow from ${
                activeTargets.length
              } to ${targetCount} paths.`
            );
            activeTargets = activeTargets.slice(0, targetCount);
          } else {
            execLogs.push(
              `    [EVOLVE] Laminar Flow (Re=${re.toFixed(
                2
              )}). Maintaining full superposition.`
            );
          }
        }

        // 2. SUPERPOSE corollary: Probabilistic Amplitude selection
        if (edge.type === 'SUPERPOSE') {
          const rawP = edge.properties.p;
          const threshold = typeof rawP === 'number' ? rawP : parseFloat(String(rawP ?? '1.0'));
          activeTargets = activeTargets.filter(
            () => Math.random() <= threshold
          );
          if (activeTargets.length === 0 && edge.targetIds[0] !== undefined) activeTargets = [edge.targetIds[0]]; // Always keep at least one
          const prunedTargets = Math.max(
            0,
            edge.targetIds.length - activeTargets.length
          );
          probabilityInputMass += prunedTargets;
          probabilityTruncatedMass += prunedTargets;
          execLogs.push(
            `    [SUPERPOSE] Amplitude p=${threshold}. Active wave-function: [${activeTargets.join(
              ', '
            )}]`
          );
        }

        // Distribution logic
        const payloads =
          Array.isArray(currentPayload) &&
          currentPayload.length === activeTargets.length
            ? currentPayload
            : activeTargets.map(() => currentPayload);

        // 3. ENTANGLE corollary: Shared Mutable State across parallel paths
        let sharedState: unknown = null;
        if (edge.type === 'ENTANGLE' || edge.properties.entangled === 'true') {
          sharedState = {
            value: currentPayload,
            timestamp: Date.now(),
            metadata: {},
          };
          execLogs.push(
            `    [ENTANGLE] Creating shared confluence state for parallel branches.`
          );
        }

        // Handle TUNNEL edges (early exits from superposition)
        // R0 rotation: use adjacency map instead of full edge scan
        let tunnelEdge: ASTEdge | undefined;
        let collapseEdge: ASTEdge | undefined;
        for (const targetId of activeTargets) {
          const targetEdges = outEdges.get(targetId.trim());
          if (!targetEdges) continue;
          for (const e of targetEdges) {
            if (!tunnelEdge && e.type === 'TUNNEL') tunnelEdge = e;
            if (
              !collapseEdge &&
              (e.type === 'RACE' || e.type === 'FOLD' || e.type === 'COLLAPSE')
            )
              collapseEdge = e;
            if (tunnelEdge && collapseEdge) break;
          }
          if (tunnelEdge && collapseEdge) break;
        }
        if (tunnelEdge) {
          execLogs.push(
            `  -> Found TUNNEL path: ${tunnelEdge.sourceIds.join('|')} -> ${
              tunnelEdge.targetIds[0]
            }`
          );
        }

        if (!collapseEdge) {
          execLogs.push(
            `Pipeline suspended in superposition. No collapse found.`
          );
          probabilityInputMass += activeTargets.length;
          probabilityUnobservedMass += activeTargets.length;
          break;
        }
        const collapseAuthorization = this.authorizeEdge(
          collapseEdge,
          currentNodeId,
          currentPayload
        );
        if (!collapseAuthorization.allowed) {
          execLogs.push(
            `  [AUTH] Denied ${collapseEdge.type}: ${collapseAuthorization.reason}`
          );
          this.tracker.updateStream(sid, 'vented');
          probabilityInputMass += activeTargets.length;
          probabilityUnobservedMass += activeTargets.length;
          break;
        }
        const concurrentResult = await this.executeConcurrentBlock({
          activeAst,
          activeTargets: activeTargets.map((targetId) => targetId.trim()),
          payloads,
          sharedState,
          collapseEdge,
          streamCounter,
          execLogs,
          cacheMetrics,
          cacheSession,
        });
        currentPayload = concurrentResult.payload;
        if (concurrentResult.outcomes) {
          const successfulBranches = concurrentResult.outcomes.filter(
            (outcome) => outcome.status === 'success'
          ).length;
          probabilityInputMass += concurrentResult.outcomes.length;
          probabilityOutputMass += successfulBranches;
          probabilityCoarseningDebt +=
            concurrentResult.outcomes.length - successfulBranches;
        } else if (concurrentResult.cacheReused) {
          probabilityInputMass += 1;
          probabilityOutputMass += 1;
        }
        this.syncExecutionAuthFromPayload(currentPayload);
        streamCounter = concurrentResult.streamCounter;
        currentNodeId = String(collapseEdge.targetIds[0] || '').trim();
        continue;
      }

      if (edge.type === 'VENT') {
        execLogs.push(`  -> VENTING path: ${edge.sourceIds[0]}`);
        this.executionBoundary.vent();
        probabilityInputMass += 1;
        probabilityCoarseningDebt += 1;
      }

      // LAMINAR: hella-whipped pipeline — self-contained fork/race/fold
      // Internally races codecs per chunk, but externally acts like PROCESS.
      // The payload passes through compressed → handler → decompressed.
      //
      // Void walking (THM-VOID-GRADIENT): the laminar edge now supports
      // adaptive codec selection via the void walker. Set voidWalking='true'
      // on the edge to enable. The walker persists across LAMINAR invocations
      // within the same engine execution, enabling cross-resource learning.
      //
      // The walker discovers content characteristics from the rejection
      // pattern alone -- no content-type sniffing needed. Mixed content
      // (JS + CSS + images) triggers automatic codec pruning after the
      // warmup period. The race still guarantees correctness (identity
      // always participates), but saves CPU by not racing codecs that
      // consistently lose.
      if (edge.type === 'LAMINAR') {
        const rawChunk = edge.properties.chunk;
        const chunkSize = typeof rawChunk === 'number' ? rawChunk : parseInt(String(rawChunk ?? '65536'), 10);
        const codecList = String(edge.properties.codecs ?? 'identity');
        const codecs = codecList.split(',').map((c: string) => c.trim());
        const voidWalking = String(edge.properties.voidWalking) !== 'false'; // default on
        execLogs.push(
          `  [LAMINAR] Hella-whipped pipeline: chunk=${chunkSize}, codecs=[${codecs.join(
            ', '
          )}], voidWalking=${voidWalking}`
        );
        // LAMINAR is β₁-neutral: internal fork/race/fold is self-contained.
        //
        // Implementation follows the paradigm (THM-TOPO-RACE-SUBSUMPTION):
        //   FORK: race all codecs on the payload
        //   RACE: select the smallest result (winner-take-all)
        //   FOLD: emit the winning codec's output
        //
        // Identity always participates (THM-TOPO-RACE-IDENTITY-BASELINE),
        // so the race can never produce a result larger than the input.
        const inputData =
          typeof currentPayload === 'string'
            ? currentPayload
            : JSON.stringify(currentPayload);
        const inputSize = inputData.length;

        // FORK: race all codecs on the input
        const codecResults: { codec: string; size: number; data: string }[] =
          [];
        for (const codec of codecs) {
          // Each codec "compresses" -- in the engine abstraction layer,
          // we measure payload size. Real compression happens in the
          // transport layer (x-gnosis/gnosis-uring). Here we track
          // which codec wins per invocation for void walking.
          codecResults.push({
            codec,
            size: codec === 'identity' ? inputSize : inputSize, // size-equal in engine layer
            data: inputData,
          });
        }

        // RACE: select smallest (THM-TOPO-RACE-SUBSUMPTION: racing <= fixed)
        const winner = codecResults.reduce((best, curr) =>
          curr.size < best.size ? curr : best
        );

        // FOLD: emit winner, VENT losers into void boundary
        const ventedCodecs = codecResults
          .filter((r) => r.codec !== winner.codec)
          .map((r) => r.codec);

        execLogs.push(
          `  [LAMINAR] RACE winner: ${winner.codec} (${
            winner.size
          }B), vented: [${ventedCodecs.join(', ')}]`
        );

        // Track codec selection in payload metadata for downstream learning
        if (typeof currentPayload === 'object' && currentPayload !== null) {
          (currentPayload as Record<string, unknown>).__laminar = {
            winner: winner.codec,
            codecs: codecs.length,
            vented: ventedCodecs,
            inputSize,
          };
        }
      }

      // Normal sequential flow
      currentNodeId = edge.targetIds[0] ? String(edge.targetIds[0]).trim() : undefined;
    }

    execLogs.push(`Final System Result: ${JSON.stringify(currentPayload)}`);
    this.executionBoundary.adapt();
    const depth = this.projectionDelayDepth;
    const projectionMetrics: GnosisEngineProjectionMetrics = {
      delayDepth: depth,
      projectedOps: this.projectionProjectedOps,
      tritStates: 3 ** depth,
      bitStates: 2 ** depth,
      informationAdvantage: 3 ** depth - 2 ** depth,
      advantageRatio: depth === 0 ? 1 : 3 ** depth / 2 ** depth,
    };
    const probabilityMassLoss =
      probabilityUnobservedMass +
      probabilityTruncatedMass +
      probabilityCoarseningDebt;
    const probabilityCertificate = createProbabilityProcessCertificate({
      certificateId: `gnosis-engine/${Date.now()}`,
      target: 'GnosisEngine.executeWithResult',
      inputMass: probabilityInputMass,
      outputMass: probabilityOutputMass,
      massLoss: probabilityMassLoss,
      residual: {
        unobservedMass: probabilityUnobservedMass,
        truncatedMass: probabilityTruncatedMass,
        coarseningDebt: probabilityCoarseningDebt,
      },
      residualBudget: probabilityInputMass,
      theoremRefs: [
        'Gnosis.FiniteProbabilityCore.RuntimeProbabilityProcessCertificate',
        'Gnosis.FiniteProbabilityCore.runtime_process_certificate_no_hidden_defect',
        'Gnosis.FiniteProbabilityCore.RuntimeTopologyProbabilityCertificate',
      ],
    });
    const probabilityTrace = this.buildRuntimeProbabilityTrace(
      probabilityCertificate,
      cacheMetrics,
      'GnosisEngine.executeWithResult',
      `gnosis-engine/trace/${Date.now()}`
    );
    return {
      logs: execLogs.join('\n'),
      payload: currentPayload,
      cacheMetrics,
      voidSnapshot: this.executionBoundary.snapshot(),
      projectionMetrics,
      decisionMetrics,
      probabilityCertificate,
      probabilityTrace,
    };
  }

  public get executionAuth(): GnosisExecutionAuthContext | null {
    return this.runtimeExecutionAuth
      ? this.cloneExecutionAuth(this.runtimeExecutionAuth)
      : null;
  }

  public get coreCache(): GnosisCoreCache | null {
    return this.runtimeCoreCache;
  }

  private async prepareCacheSession(
    ast: GraphAST,
    initialPayload: unknown,
    options: GnosisEngineExecuteOptions
  ): Promise<GnosisCoreCacheSession | null> {
    if (!this.runtimeCoreCache) {
      return null;
    }

    return this.runtimeCoreCache.createSession(
      ast,
      initialPayload,
      options.cache ?? {}
    );
  }

  private buildCacheState(
    status: GnosisEngineExecutionCacheState['status'],
    session: GnosisCoreCacheSession
  ): GnosisEngineExecutionCacheState {
    return {
      status,
      corridorKey: session.corridorKey,
      requestKey: session.requestKey,
      reuseScope: session.reuseScope,
    };
  }

  private async applyProbabilityPolicy(
    result: GnosisEngineExecutionResult,
    options: GnosisEngineExecuteOptions
  ): Promise<GnosisEngineExecutionResult> {
    const policy = options.probabilityPolicy ?? null;
    if (!policy || !result.probabilityTrace) {
      return result;
    }

    const decision = observeProbabilityRuntime(result.probabilityTrace, {
      ...(policy.residualBudget !== undefined && {
        residualBudget: policy.residualBudget,
      }),
      ...(policy.escalationBudget !== undefined && {
        escalationBudget: policy.escalationBudget,
      }),
    });
    await policy.onDecision?.(decision, result);

    const mode = policy.mode ?? 'observe';
    if (
      mode === 'fail-closed' &&
      decision.decision !== 'continue'
    ) {
      throw new Error(
        `[GnosisEngine probability] ${decision.decision}: ${decision.reason}`
      );
    }
    if (
      mode === 'cover-required' &&
      (decision.decision === 'cover' ||
        decision.decision === 'human-escalation')
    ) {
      throw new Error(
        `[GnosisEngine probability] ${decision.decision}: ${decision.reason}`
      );
    }

    return result;
  }

  private buildCachedExecutionResult(
    status: 'hit' | 'tunnel',
    payload: unknown,
    session: GnosisCoreCacheSession,
    law: LawState,
    winnerPath: string | null = null
  ): GnosisEngineExecutionResult {
    return {
      logs: this.buildCachedLogs(status, payload, session, winnerPath),
      payload,
      law,
      cache: this.buildCacheState(status, session),
      probability: createProbabilityProcessCertificate({
        certificateId: `gnosis-engine/cache/${status}/${session.requestKey}`,
        target: 'GnosisEngine.executeWithResult.cache',
        inputMass: 1,
        outputMass: 1,
        massLoss: 0,
        theoremRefs: [
          'Gnosis.FiniteProbabilityCore.RuntimeProbabilityProcessCertificate',
        ],
      }),
      probabilityTrace: createProbabilityTraceCertificate({
        certificateId: `gnosis-engine/cache-trace/${status}/${session.requestKey}`,
        target: 'GnosisEngine.executeWithResult.cache',
        mode: 'additive',
        steps: [
          createProbabilityProcessCertificate({
            certificateId: `gnosis-engine/cache/${status}/${session.requestKey}`,
            target: 'GnosisEngine.executeWithResult.cache',
            inputMass: 1,
            outputMass: 1,
            massLoss: 0,
            theoremRefs: [
              'Gnosis.FiniteProbabilityCore.RuntimeProbabilityProcessCertificate',
            ],
          }),
        ],
      }),
    };
  }

  private buildEmptyProbabilityCertificate(
    ast: GraphAST
  ): ProbabilityProcessCertificate {
    return createProbabilityProcessCertificate({
      certificateId: 'gnosis-engine/empty-graph',
      target: 'GnosisEngine.executeWithResult.empty',
      inputMass: ast.edges.length,
      outputMass: ast.edges.length,
      massLoss: 0,
      theoremRefs: [
        'Gnosis.FiniteProbabilityCore.RuntimeProbabilityProcessCertificate',
      ],
    });
  }

  private buildEmptyProbabilityTrace(ast: GraphAST): ProbabilityTraceCertificate {
    return createProbabilityTraceCertificate({
      certificateId: 'gnosis-engine/empty-graph-trace',
      target: 'GnosisEngine.executeWithResult.empty',
      mode: 'additive',
      steps: [this.buildEmptyProbabilityCertificate(ast)],
    });
  }

  private buildRuntimeProbabilityTrace(
    processCertificate: ProbabilityProcessCertificate,
    cacheMetrics: GnosisCoreCacheMetrics,
    target: string,
    certificateId: string
  ): ProbabilityTraceCertificate {
    const steps: ProbabilityProcessCertificate[] = [processCertificate];
    const cacheInputMass =
      cacheMetrics.firstSufficientCount +
      cacheMetrics.ventCount +
      cacheMetrics.repairDebt;

    if (cacheInputMass > 0) {
      const cacheMassLoss = cacheMetrics.ventCount + cacheMetrics.repairDebt;
      steps.push(
        createProbabilityProcessCertificate({
          certificateId: `${certificateId}/cache-evidence`,
          target: `${target}.cacheEvidence`,
          inputMass: cacheInputMass,
          outputMass: cacheMetrics.firstSufficientCount,
          massLoss: cacheMassLoss,
          residual: {
            truncatedMass: cacheMetrics.ventCount,
            coarseningDebt: cacheMetrics.repairDebt,
          },
          residualBudget: cacheInputMass,
          theoremRefs: [
            'Gnosis.FiniteProbabilityCore.RuntimeProbabilityProcessCertificate',
            'Gnosis.FiniteProbabilityCore.RuntimeTopologyProbabilityCertificate',
          ],
        })
      );
    }

    return createProbabilityTraceCertificate({
      certificateId,
      target,
      mode: 'additive',
      steps,
      residualBudget: steps.reduce((sum, step) => sum + step.inputMass, 0),
      theoremRefs: [
        'Gnosis.FiniteProbabilityCore.RuntimeProbabilityProcessCertificate',
        'Gnosis.FiniteProbabilityCore.runtime_process_certificate_no_hidden_defect',
        'Gnosis.FiniteProbabilityCore.RuntimeTopologyProbabilityCertificate',
      ],
    });
  }

  private buildCachedLogs(
    status: 'hit' | 'tunnel',
    payload: unknown,
    session: GnosisCoreCacheSession,
    winnerPath: string | null
  ): string {
    const lines = ['\n[Gnosis Engine Execution]'];
    const cacheLine = [
      `[CACHE] ${status.toUpperCase()}`,
      `scope=${session.reuseScope}`,
      `corridor=${session.corridorKey}`,
    ];

    if (winnerPath) {
      cacheLine.push(`winner=${winnerPath}`);
    }

    lines.push(`  ${cacheLine.join(' ')}`);
    lines.push(`Final System Result: ${JSON.stringify(payload)}`);
    return lines.join('\n');
  }

  private recordCacheMetrics(
    metrics: GnosisCoreCacheMetrics,
    update: Partial<GnosisCoreCacheMetrics>
  ): void {
    metrics.collapseCount += update.collapseCount ?? 0;
    metrics.firstSufficientCount += update.firstSufficientCount ?? 0;
    metrics.ventCount += update.ventCount ?? 0;
    metrics.repairDebt += update.repairDebt ?? 0;
    if (update.lastWinnerPath !== undefined && update.lastWinnerPath !== null) {
      metrics.lastWinnerPath = update.lastWinnerPath;
    }
  }

  private countRepairDebt(
    outcomes: readonly ConcurrentBranchOutcome[]
  ): number {
    return outcomes.filter(
      (outcome) => outcome.status === 'error' || outcome.status === 'timeout'
    ).length;
  }

  private countDiscardedBranches(
    outcomes: readonly ConcurrentBranchOutcome[],
    acceptedPath: string | null = null
  ): number {
    if (acceptedPath) {
      return outcomes.reduce(
        (count, outcome) => count + (outcome.path === acceptedPath ? 0 : 1),
        0
      );
    }

    return outcomes.filter((outcome) => outcome.status !== 'success').length;
  }

  private async prepareConcurrentCorridorSession({
    activeAst,
    activeTargets,
    collapseEdge,
    payloads,
  }: {
    activeAst: GraphAST;
    activeTargets: readonly string[];
    collapseEdge: ASTEdge;
    payloads: readonly unknown[];
  }): Promise<GnosisCoreCacheSession | null> {
    if (!this.runtimeCoreCache) {
      return null;
    }

    const rawMode = (
      collapseEdge.properties.corridorMode ??
      collapseEdge.properties.corridor_mode
    )
      ?.trim()
      .toLowerCase();
    const mode =
      rawMode === 'readonly' || rawMode === 'bypass' ? rawMode : 'readwrite';
    const rawReuseScope = (
      collapseEdge.properties.reuseScope ?? collapseEdge.properties.reuse_scope
    )
      ?.trim()
      .toLowerCase();
    const reuseScope = rawReuseScope === 'corridor' ? 'corridor' : 'exact';

    return this.runtimeCoreCache.createSession(
      this.buildConcurrentSegmentAst(activeAst, activeTargets, collapseEdge),
      payloads,
      {
        mode,
        reuseScope,
        ...(collapseEdge.properties.corridor && { corridorKey: collapseEdge.properties.corridor.trim() }),
      }
    );
  }

  private buildConcurrentSegmentAst(
    activeAst: GraphAST,
    activeTargets: readonly string[],
    collapseEdge: ASTEdge
  ): GraphAST {
    const nodes = new Map<string, ASTNode>();

    for (const targetId of activeTargets) {
      const node = this.getNode(activeAst, targetId);
      if (node) {
        nodes.set(targetId, {
          id: node.id,
          labels: [...node.labels],
          properties: { ...node.properties },
        });
      }
    }

    for (const targetId of collapseEdge.targetIds) {
      const node = this.getNode(activeAst, targetId.trim());
      if (node) {
        nodes.set(targetId.trim(), {
          id: node.id,
          labels: [...node.labels],
          properties: { ...node.properties },
        });
      }
    }

    return {
      nodes,
      edges: [
        {
          sourceIds: [...activeTargets],
          targetIds: [...collapseEdge.targetIds],
          type: collapseEdge.type,
          properties: { ...collapseEdge.properties },
        },
      ],
    };
  }

  private scheduleTargetsByCorridor(
    corridorSession: GnosisCoreCacheSession | null,
    activeTargets: readonly string[],
    payloads: readonly unknown[],
    execLogs: string[]
  ): {
    targets: string[];
    payloads: unknown[];
  } {
    if (!this.runtimeCoreCache || !corridorSession) {
      return {
        targets: [...activeTargets],
        payloads: [...payloads],
      };
    }

    const rankedTargets = this.runtimeCoreCache.rankPaths(
      corridorSession,
      activeTargets
    );
    const payloadByTarget = new Map(
      activeTargets.map((target, index) => [target, payloads[index] ?? null])
    );
    const rankedPayloads = rankedTargets.map(
      (target) => payloadByTarget.get(target) ?? null
    );
    const strength = this.runtimeCoreCache.corridorStrength(
      corridorSession.corridorKey
    );

    execLogs.push(
      `   [CORRIDOR] Ranked paths: [${rankedTargets.join(
        ', '
      )}] strength=${strength.toFixed(2)}`
    );

    return {
      targets: rankedTargets,
      payloads: rankedPayloads,
    };
  }

  private buildConcurrentCorridorObservations(
    collapseEdge: ASTEdge,
    outcomes: readonly ConcurrentBranchOutcome[]
  ): GnosisCoreCacheObservation[] {
    const raceWinnerPath =
      collapseEdge.type === 'RACE'
        ? outcomes.find((outcome) => outcome.status === 'success')?.path ?? null
        : null;

    return outcomes.map((outcome) => {
      if (
        outcome.status === 'success' &&
        (collapseEdge.type !== 'RACE' || outcome.path === raceWinnerPath)
      ) {
        return {
          path: outcome.path,
          role: 'winner',
          status: 'success',
        };
      }

      if (outcome.status === 'error' || outcome.status === 'timeout') {
        return {
          path: outcome.path,
          role: 'repair' as const,
          status: outcome.status,
          ...(outcome.reason !== undefined && { detail: outcome.reason }),
        };
      }

      return {
        path: outcome.path,
        role: 'vent' as const,
        status:
          outcome.status === 'cancelled' || outcome.status === 'vented'
            ? outcome.status
            : ('vented' as const),
        ...(outcome.reason !== undefined && { detail: outcome.reason }),
      };
    });
  }

  private async executeConcurrentBlock({
    activeAst,
    activeTargets,
    payloads,
    sharedState,
    collapseEdge,
    streamCounter,
    execLogs,
    cacheMetrics,
    cacheSession,
  }: {
    activeAst: GraphAST;
    activeTargets: string[];
    payloads: unknown[];
    sharedState: unknown;
    collapseEdge: ASTEdge;
    streamCounter: number;
    execLogs: string[];
    cacheMetrics: GnosisCoreCacheMetrics;
    cacheSession: GnosisCoreCacheSession | null;
  }): Promise<{
    payload: unknown;
    streamCounter: number;
    outcomes?: readonly ConcurrentBranchOutcome[];
    cacheReused?: boolean;
  }> {
    const corridorSession = await this.prepareConcurrentCorridorSession({
      activeAst,
      activeTargets,
      collapseEdge,
      payloads,
    });
    const effectiveCacheSession = corridorSession ?? cacheSession;
    const policy = this.parseStructuredConcurrencyPolicy(collapseEdge);
    this.logStructuredPolicy(collapseEdge, policy, execLogs);

    await this.notifyEdgeEvaluated(collapseEdge);

    if (
      corridorSession &&
      this.runtimeCoreCache &&
      corridorSession.mode !== 'bypass'
    ) {
      const lookup = this.runtimeCoreCache.lookup<unknown>(corridorSession);
      if (lookup.kind === 'hit') {
        const cached = this.runtimeCoreCache.readPayload(lookup.entry);
        if (cached.found) {
          execLogs.push(
            `   [CORRIDOR] HIT corridor=${corridorSession.corridorKey} winner=${
              lookup.entry.winnerPath ?? 'cached'
            }`
          );
          this.recordCacheMetrics(cacheMetrics, {
            collapseCount: 1,
            firstSufficientCount: 1,
            lastWinnerPath: lookup.entry.winnerPath ?? null,
          });
          return {
            payload: cached.payload,
            streamCounter,
            cacheReused: true,
          };
        }
      }

      if (lookup.kind === 'tunnel') {
        execLogs.push(
          `   [CORRIDOR] TUNNEL corridor=${corridorSession.corridorKey}`
        );
        const payload = await lookup.promise;
        this.recordCacheMetrics(cacheMetrics, {
          collapseCount: 1,
          firstSufficientCount: 1,
        });
        return {
          payload,
          streamCounter,
          cacheReused: true,
        };
      }

      execLogs.push(
        `   [CORRIDOR] MISS corridor=${corridorSession.corridorKey}`
      );
    }

    const scheduled = this.scheduleTargetsByCorridor(
      corridorSession,
      activeTargets,
      payloads,
      execLogs
    );
    const executeFresh = async (): Promise<{
      payload: unknown;
      streamCounter: number;
      outcomes: ConcurrentBranchOutcome[];
      blockMetrics: GnosisCoreCacheMetrics;
    }> => {
      // ─── Frame-native fast path ───────────────────────────────────────
      // When there are no timeouts, no deadlines, no shared state, and all
      // handlers exist, bypass AbortController/ConcurrentBranchHandle overhead
      // entirely. Direct Promise.race/allSettled on raw handler calls.
      //
      // This gives all .gg topology executions the same 4-5x speedup that
      // frame-native fold/race achieves in aeon-pipelines benchmarks.
      if (
        this.canUseFrameNativePath(
          activeAst,
          scheduled.targets,
          policy,
          sharedState
        )
      ) {
        const frameResult = await this.executeFrameNative({
          activeAst,
          activeTargets: scheduled.targets,
          payloads: scheduled.payloads,
          collapseEdge,
          streamCounter,
          execLogs,
          cacheMetrics,
          cacheSession: effectiveCacheSession,
        });
        if (frameResult !== null) {
          if (corridorSession && this.runtimeCoreCache) {
            this.runtimeCoreCache.recordEvidence(
              corridorSession,
              this.buildConcurrentCorridorObservations(
                collapseEdge,
                frameResult.outcomes
              )
            );
            execLogs.push(
              `   [CORRIDOR] Updated strength=${this.runtimeCoreCache
                .corridorStrength(corridorSession.corridorKey)
                .toFixed(2)}`
            );
          }
          return {
            payload: frameResult.payload,
            streamCounter: frameResult.streamCounter,
            outcomes: frameResult.outcomes,
            blockMetrics: this.buildConcurrentBlockMetrics(
              collapseEdge,
              frameResult.outcomes
            ),
          };
        }
        // Fall through to standard path if frame-native fails
      }

      const blockStartedAt = Date.now();
      const deadlineAt =
        policy.deadlineMs === null ? null : blockStartedAt + policy.deadlineMs;

      const branches = scheduled.targets.map((targetId, index) => {
        const sid = streamCounter++;
        this.tracker.updateStream(sid, 'active');

        return this.startConcurrentBranch({
          activeAst,
          targetId,
          payload: scheduled.payloads[index] ?? null,
          sharedState,
          sid,
          timeoutMs: this.resolveConcurrentTimeout(
            policy.timeoutMs,
            deadlineAt
          ),
          cacheSession: effectiveCacheSession,
        });
      });

      const result =
        collapseEdge.type === 'RACE'
          ? await this.resolveRaceCollapse(
              branches,
              collapseEdge,
              policy,
              execLogs,
              cacheMetrics
            )
          : await this.resolveFoldCollapse(
              branches,
              collapseEdge,
              policy,
              execLogs,
              cacheMetrics
            );

      this.logBranchOutcomes(result.outcomes, execLogs);
      if (corridorSession && this.runtimeCoreCache) {
        this.runtimeCoreCache.recordEvidence(
          corridorSession,
          this.buildConcurrentCorridorObservations(
            collapseEdge,
            result.outcomes
          )
        );
        execLogs.push(
          `   [CORRIDOR] Updated strength=${this.runtimeCoreCache
            .corridorStrength(corridorSession.corridorKey)
            .toFixed(2)}`
        );
      }

      return {
        payload: result.payload,
        streamCounter,
        outcomes: result.outcomes,
        blockMetrics: this.buildConcurrentBlockMetrics(
          collapseEdge,
          result.outcomes
        ),
      };
    };

    if (
      corridorSession &&
      this.runtimeCoreCache &&
      corridorSession.mode === 'readwrite'
    ) {
      const freshExecution = executeFresh();
      this.runtimeCoreCache.registerInflight(
        corridorSession,
        freshExecution.then((result) => result.payload)
      );

      try {
        const result = await freshExecution;
        this.runtimeCoreCache.commit(
          corridorSession,
          result.payload,
          result.blockMetrics
        );
        return {
          payload: result.payload,
          streamCounter: result.streamCounter,
          outcomes: result.outcomes,
        };
      } catch (error) {
        this.runtimeCoreCache.fail(corridorSession, error);
        throw error;
      } finally {
        this.runtimeCoreCache.release(corridorSession);
      }
    }

    const result = await executeFresh();
    return {
      payload: result.payload,
      streamCounter: result.streamCounter,
      outcomes: result.outcomes,
    };
  }

  /**
   * Check if the frame-native fast path can be used.
   *
   * Conditions:
   *   1. No timeout or deadline (no AbortController needed)
   *   2. No shared state (no ENTANGLE semantics)
   *   3. All target nodes have registered handlers
   *   4. Default failure policy (cancel) — frame-native handles this naturally
   */
  private canUseFrameNativePath(
    activeAst: GraphAST,
    activeTargets: string[],
    policy: StructuredConcurrencyPolicy,
    sharedState: unknown
  ): boolean {
    // Timeout/deadline requires AbortController + setTimeout
    if (policy.timeoutMs !== null || policy.deadlineMs !== null) return false;

    // Shared state requires GnosisHandlerContext propagation
    if (sharedState !== null) return false;

    // All handlers must exist
    for (const targetId of activeTargets) {
      const node = this.getNode(activeAst, targetId);
      if (!node || !this.findHandler(node)) return false;
    }

    return true;
  }

  /**
   * Frame-native execution: bypass AbortController/ConcurrentBranchHandle entirely.
   *
   * Extracts handler functions from the registry, wraps them as work functions
   * with their payloads bound, then dispatches through frameRace or frameFold.
   *
   * Returns null if execution fails (caller falls through to standard path).
   */
  private async executeFrameNative({
    activeAst,
    activeTargets,
    payloads,
    collapseEdge,
    streamCounter,
    execLogs,
    cacheMetrics,
    cacheSession,
  }: {
    activeAst: GraphAST;
    activeTargets: string[];
    payloads: unknown[];
    collapseEdge: ASTEdge;
    streamCounter: number;
    execLogs: string[];
    cacheMetrics: GnosisCoreCacheMetrics;
    cacheSession: GnosisCoreCacheSession | null;
  }): Promise<{
    payload: unknown;
    streamCounter: number;
    outcomes: ConcurrentBranchOutcome[];
  } | null> {
    try {
      // Build raw work functions from handlers + payloads
      const workFns: (() => Promise<unknown>)[] = activeTargets.map(
        (targetId, index) => {
          const node = this.getNode(activeAst, targetId)!;
          const handler = this.findHandler(node)!;
          const payload = payloads[index] ?? null;
          return () =>
            Promise.resolve(
              handler(payload, node.properties, {
                nodeId: targetId,
                ...(this.executionAuth && { executionAuth: this.executionAuth }),
                ...(this.runtimeCoreCache && { coreCache: this.runtimeCoreCache }),
                ...(cacheSession && { cacheSession }),
                ...(this.activeLawState && { law: this.activeLawState }),
              })
            );
        }
      );

      // Update tracker for all branches
      for (let i = 0; i < activeTargets.length; i++) {
        this.tracker.updateStream(streamCounter + i, 'active');
      }

      let payload: unknown;
      const outcomes: ConcurrentBranchOutcome[] = [];

      if (collapseEdge.type === 'RACE') {
        execLogs.push(`   [FRAME-NATIVE] Racing ${activeTargets.length} paths`);
        const raceResult = await frameRace(workFns);
        payload = raceResult.result;
        const winnerTarget = activeTargets[raceResult.winnerIndex];
        execLogs.push(`   Race concluded! Winner: ${winnerTarget}`);
        this.recordCacheMetrics(cacheMetrics, {
          collapseCount: 1,
          firstSufficientCount: 1,
          ventCount: Math.max(0, activeTargets.length - 1),
          ...(winnerTarget !== undefined && { lastWinnerPath: winnerTarget }),
        });

        // Update tracker and void boundary
        for (let i = 0; i < activeTargets.length; i++) {
          const sid = streamCounter + i;
          const status = i === raceResult.winnerIndex ? 'success' : 'vented';
          this.tracker.updateStream(
            sid,
            i === raceResult.winnerIndex ? 'completed' : 'vented'
          );
          if (i !== raceResult.winnerIndex) {
            this.executionBoundary.raceLoser(i);
            this.executionBoundary.recordExecution(i, true);
          }
          outcomes.push({
            path: activeTargets[i]!,
            sid,
            status,
            durationMs: 0,
            ...(i === raceResult.winnerIndex ? { value: payload } : {}),
          });
        }
      } else {
        // FOLD / COLLAPSE
        execLogs.push(
          `   [FRAME-NATIVE] Folding ${activeTargets.length} paths`
        );
        const results = await Promise.allSettled(workFns.map((fn) => fn()));
        const successValues: unknown[] = [];
        const successTargets: string[] = [];

        for (let i = 0; i < results.length; i++) {
          const sid = streamCounter + i;
          const res = results[i];
          const at = activeTargets[i];
          if (!res || at === undefined) continue;
          if (res.status === 'fulfilled') {
            const value = (res as PromiseFulfilledResult<unknown>).value;
            successValues.push(value);
            successTargets.push(at);
            this.tracker.updateStream(sid, 'completed');
            outcomes.push({
              path: at,
              sid,
              status: 'success',
              durationMs: 0,
              value,
            });
          } else {
            this.tracker.updateStream(sid, 'vented');
            outcomes.push({
              path: at,
              sid,
              status: 'error',
              durationMs: 0,
              reason: this.errorMessage(
                (res as PromiseRejectedResult).reason
              ),
            });
          }
        }

        if (successValues.length === 0) {
          return null; // All failed — fall back to standard path for error handling
        }

        // Merge using the same logic as the standard fold
        payload = this.mergeFrameNativeFoldResults(
          successValues,
          successTargets
        );
        execLogs.push(
          `   Folded result: ${JSON.stringify(payload).substring(0, 50)}...`
        );
        this.recordCacheMetrics(cacheMetrics, {
          collapseCount: 1,
          firstSufficientCount: 1,
          ventCount: results.length - successValues.length,
          repairDebt: results.length - successValues.length,
          lastWinnerPath:
            collapseEdge.targetIds[0]?.trim() ?? successTargets[0] ?? null,
        });
      }

      this.syncExecutionAuthFromPayload(payload);
      return {
        payload,
        streamCounter: streamCounter + activeTargets.length,
        outcomes,
      };
    } catch {
      return null; // Frame-native failed — fall through to standard path
    }
  }

  private buildConcurrentBlockMetrics(
    collapseEdge: ASTEdge,
    outcomes: readonly ConcurrentBranchOutcome[]
  ): GnosisCoreCacheMetrics {
    const acceptedPath =
      collapseEdge.type === 'RACE'
        ? outcomes.find((outcome) => outcome.status === 'success')?.path ?? null
        : null;

    return {
      collapseCount: 1,
      firstSufficientCount: outcomes.some(
        (outcome) => outcome.status === 'success'
      )
        ? 1
        : 0,
      ventCount: this.countDiscardedBranches(outcomes, acceptedPath),
      repairDebt: this.countRepairDebt(outcomes),
      lastWinnerPath:
        acceptedPath ??
        collapseEdge.targetIds[0]?.trim() ??
        outcomes.find((outcome) => outcome.status === 'success')?.path ??
        null,
    };
  }

  /**
   * Merge fold results from frame-native execution.
   * Mirrors the logic in `mergeFoldOutcomes` but without ConcurrentBranchOutcome wrapping.
   */
  private mergeFrameNativeFoldResults(
    values: unknown[],
    targets: string[]
  ): unknown {
    // Numeric vector sum (same as standard fold)
    if (
      values.length > 0 &&
      values.every(
        (v) =>
          Array.isArray(v) &&
          (v as unknown[]).every((e) => typeof e === 'number')
      )
    ) {
      const vectors = values as number[][];
      return vectors[0]!.map((_, i) =>
        vectors.reduce((sum, vec) => sum + (vec[i] ?? 0), 0)
      );
    }

    // Object merge keyed by target path
    const merged: Record<string, unknown> = {};
    for (let i = 0; i < values.length; i++) {
      const t = targets[i];
      if (t === undefined) continue;
      merged[t] = values[i];
    }
    return merged;
  }

  private startConcurrentBranch({
    activeAst,
    targetId,
    payload,
    sharedState,
    sid,
    timeoutMs,
    cacheSession,
  }: {
    activeAst: GraphAST;
    targetId: string;
    payload: unknown;
    sharedState: unknown;
    sid: number;
    timeoutMs: number | null;
    cacheSession: GnosisCoreCacheSession | null;
  }): ConcurrentBranchHandle {
    const controller = new AbortController();
    const branch: ConcurrentBranchHandle = {
      path: targetId,
      sid,
      controller,
      settled: false,
      promise: Promise.resolve({
        path: targetId,
        sid,
        status: 'vented',
        durationMs: 0,
        reason: 'Branch not started.',
      }),
    };

    branch.promise = this.invokeConcurrentBranch({
      activeAst,
      targetId,
      payload,
      sharedState,
      sid,
      controller,
      timeoutMs,
      cacheSession,
    }).then((outcome) => {
      branch.settled = true;
      branch.outcome = outcome;
      return outcome;
    });

    return branch;
  }

  private async invokeConcurrentBranch({
    activeAst,
    targetId,
    payload,
    sharedState,
    sid,
    controller,
    timeoutMs,
    cacheSession,
  }: {
    activeAst: GraphAST;
    targetId: string;
    payload: unknown;
    sharedState: unknown;
    sid: number;
    controller: AbortController;
    timeoutMs: number | null;
    cacheSession: GnosisCoreCacheSession | null;
  }): Promise<ConcurrentBranchOutcome> {
    const startedAt = Date.now();
    const finish = (
      status: StructuredBranchStatus,
      value?: unknown,
      reason?: string
    ): ConcurrentBranchOutcome => ({
      path: targetId,
      sid,
      status,
      durationMs: Date.now() - startedAt,
      value,
      ...(reason !== undefined && { reason }),
    });

    const node = this.getNode(activeAst, targetId);
    if (!node) {
      this.tracker.updateStream(sid, 'vented');
      return finish('vented', undefined, `Node '${targetId}' not found.`);
    }

    const handler = this.findHandler(node);
    if (!handler) {
      this.tracker.updateStream(sid, 'vented');
      return finish(
        'vented',
        undefined,
        `No handler registered for '${targetId}'.`
      );
    }

    const context: GnosisHandlerContext = {
      nodeId: targetId,
      sharedState,
      signal: controller.signal,
      ...(this.executionAuth && { executionAuth: this.executionAuth }),
      ...(this.runtimeCoreCache && { coreCache: this.runtimeCoreCache }),
      ...(cacheSession && { cacheSession }),
      ...(this.activeLawState && { law: this.activeLawState }),
    };
    const handlerPromise = Promise.resolve(
      handler(payload, node.properties, context)
    );
    handlerPromise.catch(() => undefined);

    const abortPromise = new Promise<never>((_, reject) => {
      controller.signal.addEventListener(
        'abort',
        () => reject(controller.signal.reason),
        { once: true }
      );
    });

    const timeoutId =
      timeoutMs === null
        ? null
        : setTimeout(() => {
            controller.abort({
              kind: 'timeout',
              reason: `timed out after ${timeoutMs}ms.`,
            });
          }, timeoutMs);

    try {
      const result = await Promise.race([handlerPromise, abortPromise]);
      if (timeoutId !== null) {
        clearTimeout(timeoutId);
      }

      this.syncExecutionAuthFromPayload(result);
      this.tracker.updateStream(sid, 'completed');
      return finish('success', result);
    } catch (error) {
      if (timeoutId !== null) {
        clearTimeout(timeoutId);
      }

      if (controller.signal.aborted) {
        const abortReason = this.normalizeAbortReason(controller.signal.reason);
        this.tracker.updateStream(sid, 'vented');
        return finish(abortReason.status, undefined, abortReason.reason);
      }

      this.tracker.updateStream(sid, 'vented');
      return finish('error', undefined, this.errorMessage(error));
    }
  }

  private async resolveRaceCollapse(
    branches: ConcurrentBranchHandle[],
    collapseEdge: ASTEdge,
    policy: StructuredConcurrencyPolicy,
    execLogs: string[],
    cacheMetrics: GnosisCoreCacheMetrics
  ): Promise<{ payload: unknown; outcomes: ConcurrentBranchOutcome[] }> {
    execLogs.push(`   Racing paths: [${collapseEdge.sourceIds.join(', ')}]`);
    const resolution = await resolveStructuredRace(branches, policy);
    if (resolution.cancelledByFailure) {
      execLogs.push(
        `   [STRUCTURED] Race cancelled by ${resolution.cancelledByFailure.path} (${resolution.cancelledByFailure.status}).`
      );
      this.recordCacheMetrics(cacheMetrics, {
        collapseCount: 1,
        firstSufficientCount: 0,
        ventCount: this.countDiscardedBranches(resolution.outcomes),
        repairDebt: this.countRepairDebt(resolution.outcomes),
      });
      return {
        payload: this.buildConcurrentFailurePayload(
          collapseEdge.type,
          policy,
          resolution.outcomes,
          resolution.cancelledByFailure
        ),
        outcomes: [...resolution.outcomes],
      };
    }

    if (resolution.winner) {
      execLogs.push(`   Race concluded! Winner: ${resolution.winner.path}`);
      // Record losers in void boundary
      for (let i = 0; i < resolution.outcomes.length; i++) {
        if (resolution.outcomes[i]!.path !== resolution.winner.path) {
          this.executionBoundary.raceLoser(i);
          this.executionBoundary.recordExecution(i, true);
        }
      }
      this.recordCacheMetrics(cacheMetrics, {
        collapseCount: 1,
        firstSufficientCount: 1,
        ventCount: this.countDiscardedBranches(
          resolution.outcomes,
          resolution.winner.path
        ),
        repairDebt: this.countRepairDebt(resolution.outcomes),
        lastWinnerPath: resolution.winner.path,
      });
      return {
        payload: resolution.winner.value,
        outcomes: [...resolution.outcomes],
      };
    }

    execLogs.push(
      `   [STRUCTURED] Race exhausted without a successful branch.`
    );
    const primaryFailure = resolution.outcomes.find(
      (outcome) => outcome.status !== 'success'
    );
    this.recordCacheMetrics(cacheMetrics, {
      collapseCount: 1,
      firstSufficientCount: 0,
      ventCount: this.countDiscardedBranches(resolution.outcomes),
      repairDebt: this.countRepairDebt(resolution.outcomes),
    });
    return {
      payload: this.buildConcurrentFailurePayload(
        collapseEdge.type,
        policy,
        [...resolution.outcomes],
        primaryFailure
      ),
      outcomes: [...resolution.outcomes],
    };
  }

  private async resolveFoldCollapse(
    branches: ConcurrentBranchHandle[],
    collapseEdge: ASTEdge,
    policy: StructuredConcurrencyPolicy,
    execLogs: string[],
    cacheMetrics: GnosisCoreCacheMetrics
  ): Promise<{ payload: unknown; outcomes: ConcurrentBranchOutcome[] }> {
    execLogs.push(`   Folding paths: [${collapseEdge.sourceIds.join(', ')}]`);
    const resolution = await resolveStructuredFold(branches, policy);
    if (resolution.cancelledByFailure) {
      execLogs.push(
        `   [STRUCTURED] Fold cancelled by ${resolution.cancelledByFailure.path} (${resolution.cancelledByFailure.status}).`
      );
      this.recordCacheMetrics(cacheMetrics, {
        collapseCount: 1,
        firstSufficientCount: 0,
        ventCount: this.countDiscardedBranches(resolution.outcomes),
        repairDebt: this.countRepairDebt(resolution.outcomes),
      });
      return {
        payload: this.buildConcurrentFailurePayload(
          collapseEdge.type,
          policy,
          resolution.outcomes,
          resolution.cancelledByFailure
        ),
        outcomes: [...resolution.outcomes],
      };
    }

    if (!resolution.outcomes.some((outcome) => outcome.status === 'success')) {
      execLogs.push(`   [STRUCTURED] Fold produced no surviving branches.`);
      this.recordCacheMetrics(cacheMetrics, {
        collapseCount: 1,
        firstSufficientCount: 0,
        ventCount: this.countDiscardedBranches(resolution.outcomes),
        repairDebt: this.countRepairDebt(resolution.outcomes),
      });
      return {
        payload: this.buildConcurrentFailurePayload(
          collapseEdge.type,
          policy,
          [...resolution.outcomes],
          resolution.outcomes[0]
        ),
        outcomes: [...resolution.outcomes],
      };
    }

    const payload = this.mergeFoldOutcomes(
      [...resolution.outcomes],
      policy,
      collapseEdge
    );
    // FOLD: merge all successful branch boundaries (each contributes a vent for failed branches)
    for (const outcome of resolution.outcomes) {
      if (outcome.status !== 'success') {
        this.executionBoundary.vent();
      }
    }
    execLogs.push(
      `   Folded result: ${JSON.stringify(payload).substring(0, 50)}...`
    );
    this.recordCacheMetrics(cacheMetrics, {
      collapseCount: 1,
      firstSufficientCount: 1,
      ventCount: this.countDiscardedBranches(resolution.outcomes),
      repairDebt: this.countRepairDebt(resolution.outcomes),
      lastWinnerPath:
        collapseEdge.targetIds[0]?.trim() ??
        resolution.outcomes.find((outcome) => outcome.status === 'success')
          ?.path ??
        null,
    });
    return { payload, outcomes: [...resolution.outcomes] };
  }

  private mergeFoldOutcomes(
    outcomes: readonly ConcurrentBranchOutcome[],
    policy: StructuredConcurrencyPolicy,
    collapseEdge?: ASTEdge
  ): unknown {
    // Check for registered fold strategy (e.g., merge-ast)
    const strategyName = collapseEdge?.properties?.strategy;
    if (strategyName) {
      const strategyHandler = this.registry.getHandler(`fold:${strategyName}`);
      if (strategyHandler) {
        // Synchronous dispatch: build keyed results map and call handler
        const results = new Map<string, unknown>();
        for (const outcome of outcomes) {
          if (outcome.status === 'success') {
            results.set(outcome.path, outcome.value);
          }
        }
        // Return synchronously -- strategy handlers for fold are expected to be sync-compatible
        return strategyHandler(results, collapseEdge?.properties ?? {});
      }
    }

    const successfulValues = outcomes
      .filter((outcome) => outcome.status === 'success')
      .map((outcome) => outcome.value);

    if (
      successfulValues.length > 0 &&
      successfulValues.every(
        (value) =>
          Array.isArray(value) &&
          value.every((entry) => typeof entry === 'number')
      )
    ) {
      const numericVectors = successfulValues as number[][];
      return numericVectors[0]!.map((_, index) =>
        numericVectors.reduce((sum, vector) => sum + (vector[index] ?? 0), 0)
      );
    }

    const merged: Record<string, unknown> = {};
    for (const outcome of outcomes) {
      if (outcome.status === 'success') {
        merged[outcome.path] = outcome.value;
        continue;
      }

      if (policy.failure === 'shield') {
        merged[outcome.path] = this.encodeBranchOutcome(outcome);
      }
    }

    return merged;
  }

  private encodeBranchOutcome(
    outcome: ConcurrentBranchOutcome
  ): Record<string, unknown> {
    return {
      kind: outcome.status,
      path: outcome.path,
      durationMs: outcome.durationMs,
      ...(outcome.reason ? { reason: outcome.reason } : {}),
    };
  }

  private buildConcurrentFailurePayload(
    collapseType: string,
    policy: StructuredConcurrencyPolicy,
    outcomes: readonly ConcurrentBranchOutcome[],
    primaryFailure?: ConcurrentBranchOutcome
  ): { kind: 'err'; error: Record<string, unknown> } {
    return {
      kind: 'err',
      error: {
        type: 'structured-concurrency',
        collapse: collapseType,
        failureMode: policy.failure,
        branch: primaryFailure?.path,
        status: primaryFailure?.status ?? 'error',
        reason:
          primaryFailure?.reason ?? 'No successful branch completed in time.',
        branches: outcomes.map((outcome) => this.encodeBranchOutcome(outcome)),
      },
    };
  }

  private logStructuredPolicy(
    collapseEdge: ASTEdge,
    policy: StructuredConcurrencyPolicy,
    execLogs: string[]
  ): void {
    const policyParts = [`failure=${policy.failure}`];
    if (policy.timeoutMs !== null) {
      policyParts.push(`timeoutMs=${policy.timeoutMs}`);
    }
    if (policy.deadlineMs !== null) {
      policyParts.push(`deadlineMs=${policy.deadlineMs}`);
    }

    execLogs.push(
      `   [STRUCTURED] ${collapseEdge.type} policy: ${policyParts.join(', ')}`
    );
  }

  private logBranchOutcomes(
    outcomes: ConcurrentBranchOutcome[],
    execLogs: string[]
  ): void {
    for (const outcome of outcomes) {
      execLogs.push(
        `   [BRANCH ${outcome.path}] ${outcome.status} (${
          outcome.durationMs
        }ms)${outcome.reason ? `: ${outcome.reason}` : ''}`
      );
    }
  }

  private parseStructuredConcurrencyPolicy(
    edge: ASTEdge
  ): StructuredConcurrencyPolicy {
    const rawFailure = edge.properties.failure?.trim().toLowerCase();
    const failure: StructuredConcurrencyFailureMode =
      rawFailure === 'vent' || rawFailure === 'shield' ? rawFailure : 'cancel';

    return {
      failure,
      timeoutMs: this.parsePositiveInteger(
        edge.properties.timeoutMs ?? edge.properties.timeout
      ),
      deadlineMs: this.parsePositiveInteger(
        edge.properties.deadlineMs ?? edge.properties.deadline
      ),
    };
  }

  private resolveConcurrentTimeout(
    timeoutMs: number | null,
    deadlineAt: number | null
  ): number | null {
    const budgets: number[] = [];

    if (timeoutMs !== null) {
      budgets.push(timeoutMs);
    }
    if (deadlineAt !== null) {
      budgets.push(Math.max(0, deadlineAt - Date.now()));
    }

    if (budgets.length === 0) {
      return null;
    }

    return Math.max(0, Math.min(...budgets));
  }

  private parsePositiveInteger(raw: string | undefined): number | null {
    if (!raw) {
      return null;
    }

    const parsed = Number(raw);
    if (!Number.isFinite(parsed) || parsed < 0) {
      return null;
    }

    return Math.round(parsed);
  }

  private normalizeAbortReason(reason: unknown): {
    status: 'timeout' | 'cancelled';
    reason: string;
  } {
    if (typeof reason === 'object' && reason !== null) {
      const record = reason as Record<string, unknown>;
      const status = record.kind === 'timeout' ? 'timeout' : 'cancelled';
      return {
        status,
        reason:
          typeof record.reason === 'string'
            ? record.reason
            : status === 'timeout'
            ? 'timed out.'
            : 'cancelled.',
      };
    }

    return {
      status: 'cancelled',
      reason:
        typeof reason === 'string' && reason.length > 0 ? reason : 'cancelled.',
    };
  }

  private errorMessage(error: unknown): string {
    if (error instanceof Error) {
      return error.message;
    }
    return String(error);
  }

  private cloneExecutionAuth(
    auth: GnosisExecutionAuthContext
  ): GnosisExecutionAuthContext {
    return {
      enforce: auth.enforce === true,
      ...(auth.principal !== undefined && { principal: auth.principal }),
      ...(auth.token !== undefined && { token: auth.token }),
      capabilities: auth.capabilities.map((capability) => ({
        ...capability,
        ...(capability.constraints && { constraints: { ...capability.constraints } }),
      })),
      reactorHandleGrants: (auth.reactorHandleGrants ?? []).map((grant) => ({
        ...grant,
        claims: {
          ...grant.claims,
          operations: [...grant.claims.operations],
          ucanCapabilities: [...grant.claims.ucanCapabilities],
        },
      })),
    };
  }

  private initializeExecutionAuthState(
    initialPayload: unknown,
    executionAuth: GnosisExecutionAuthContext | null
  ): void {
    this.runtimeExecutionAuth = null;
    this.syncExecutionAuthFromPayload(initialPayload);
    if (executionAuth) {
      this.adoptExecutionAuth(executionAuth);
    }
  }

  private adoptExecutionAuth(candidate: unknown): void {
    const normalized = normalizeExecutionAuthContext(candidate);
    if (!normalized) {
      return;
    }

    if (!this.runtimeExecutionAuth) {
      this.runtimeExecutionAuth = this.cloneExecutionAuth(normalized);
      return;
    }

    this.runtimeExecutionAuth = this.mergeExecutionAuth(
      this.runtimeExecutionAuth,
      normalized
    );
  }

  private mergeExecutionAuth(
    current: GnosisExecutionAuthContext,
    incoming: GnosisExecutionAuthContext
  ): GnosisExecutionAuthContext {
    return mergeExecutionAuthContexts(current, incoming);
  }

  private findHandler(node: ASTNode): GnosisHandler | null {
    if (this.handlerCache.has(node)) {
      return this.handlerCache.get(node)!;
    }

    for (const label of node.labels) {
      const handler = this.registry.getHandler(label);
      if (handler) {
        this.handlerCache.set(node, handler);
        return handler;
      }
    }
    this.handlerCache.set(node, null);
    return null;
  }

  private extractExecutionAuth(
    payload: unknown
  ): GnosisExecutionAuthContext | null {
    if (typeof payload !== 'object' || payload === null) {
      return null;
    }

    const record = payload as Record<string, unknown>;
    return normalizeExecutionAuthContext(record.executionAuth);
  }

  private syncExecutionAuthFromPayload(payload: unknown): void {
    const executionAuth = this.extractExecutionAuth(payload);
    if (!executionAuth) {
      return;
    }

    this.adoptExecutionAuth(executionAuth);
  }

  private authorizeEdge(
    edge: ASTEdge,
    currentNodeId: string,
    payload: unknown
  ): { allowed: boolean; reason?: string } {
    this.syncExecutionAuthFromPayload(payload);
    const executionAuth = this.runtimeExecutionAuth;
    if (!executionAuth || executionAuth.enforce !== true) {
      return { allowed: true };
    }

    const sourceId = edge.sourceIds[0]?.trim() || currentNodeId;
    const targetIds = edge.targetIds.map((targetId) => targetId.trim());

    return authorizeTopologyEdge({
      edgeType: edge.type,
      sourceId,
      targetIds,
      auth: executionAuth,
    });
  }

  private async notifyEdgeEvaluated(edge: ASTEdge): Promise<void> {
    if (!this.onEdgeEvaluated) {
      return;
    }
    await this.onEdgeEvaluated(edge);
  }

  private selectEdge(edges: ASTEdge[], payload: unknown): ASTEdge {
    const matchedEdges = this.filterEdgesByPayloadCase(edges, payload);
    const candidates = matchedEdges.length > 0 ? matchedEdges : edges;

    return (
      candidates.find((edge) =>
        [
          'FORK',
          'RACE',
          'FOLD',
          'COLLAPSE',
          'EVOLVE',
          'SUPERPOSE',
          'ENTANGLE',
          'OBSERVE',
          'LAMINAR',
        ].includes(edge.type || '')
      ) || candidates[0]!
    );
  }

  private filterEdgesByPayloadCase(
    edges: ASTEdge[],
    payload: unknown
  ): ASTEdge[] {
    const caseAwareEdges = edges.filter((edge) =>
      this.edgeHasCaseConstraint(edge)
    );
    if (caseAwareEdges.length === 0) {
      return [];
    }

    const payloadTokens = this.extractPayloadCaseTokens(payload);
    if (payloadTokens.length === 0) {
      return [];
    }

    return caseAwareEdges.filter((edge) => {
      const expectedTokens = this.extractEdgeCaseTokens(edge);
      return expectedTokens.some((token) => payloadTokens.includes(token));
    });
  }

  private edgeHasCaseConstraint(edge: ASTEdge): boolean {
    return this.extractEdgeCaseTokens(edge).length > 0;
  }

  private extractEdgeCaseTokens(edge: ASTEdge): string[] {
    const fields = [
      'case',
      'match',
      'when',
      'variant',
      'status',
      'kind',
    ] as const;
    const tokens: string[] = [];

    for (const field of fields) {
      const rawValue = edge.properties[field];
      if (!rawValue) {
        continue;
      }

      const parsedTokens = rawValue
        .split(/[\s,|]+/)
        .map((value) => value.trim().toLowerCase())
        .filter((value) => value.length > 0);

      tokens.push(...parsedTokens);
    }

    return [...new Set(tokens)];
  }

  private extractPayloadCaseTokens(payload: unknown): string[] {
    if (
      typeof payload === 'string' ||
      typeof payload === 'number' ||
      typeof payload === 'boolean'
    ) {
      return [String(payload).trim().toLowerCase()];
    }

    if (
      typeof payload !== 'object' ||
      payload === null ||
      Array.isArray(payload)
    ) {
      return [];
    }

    const record = payload as Record<string, unknown>;
    const fields = ['kind', 'variant', 'status', 'case', 'type'] as const;
    const tokens: string[] = [];

    for (const field of fields) {
      const value = record[field];
      if (
        typeof value === 'string' ||
        typeof value === 'number' ||
        typeof value === 'boolean'
      ) {
        tokens.push(String(value).trim().toLowerCase());
      }
    }

    return [...new Set(tokens)];
  }

  // ─── Packed Knot Execution ───────────────────────────────────────

  /**
   * Execute a topology from a packed knot binary.
   * The knot stream is pre-toposorted and Reidemeister-optimized.
   * Two bits per crossing. No parse. Shift and mask.
   *
   * Self-modifying: VENT crossings feed output back as new crossings.
   */
  public async executeFromKnot(
    knot: import('./packed-knot.js').PackedKnot,
    initialPayload: any = null,
    options: GnosisEngineExecuteOptions = {}
  ): Promise<GnosisEngineExecutionResult> {
    const {
      release,
      FORK,
      RACE,
      FOLD,
      VENT,
      edgeTypeToCrossing,
      crossingName,
    } = await import('./packed-knot.js');

    const logs: string[] = [];
    let payload = initialPayload;
    const forkResults: any[][] = [];
    const authContext = normalizeExecutionAuthContext(
      options.executionAuth ?? options.authContext ?? null
    );
    if (authContext) {
      this.adoptExecutionAuth(authContext);
    }

    const gen = release(knot);
    let result = gen.next();

    while (!result.done) {
      const yielded = result.value;
      if (!yielded) {
        result = gen.next();
        continue;
      }

      const { op, nodeId, index } = yielded;
      const handler = this.registry.getHandler(nodeId);
      const label = crossingName(op);

      logs.push(`[knot:${index}] ${label} ${nodeId}`);

      switch (op) {
        case FORK: {
          // Fork: run handler, push result to fork stack
          if (handler) {
            const ctx: GnosisHandlerContext = {
              nodeId,
              labels: [nodeId],
              properties: {},
              edges: [],
              engineRef: this,
              ...(authContext && { authContext, executionAuth: authContext }),
            };
            const forkResult = await handler(payload, {}, ctx);
            forkResults.push(
              Array.isArray(forkResult) ? forkResult : [forkResult]
            );
          } else {
            forkResults.push([payload]);
          }
          break;
        }

        case RACE: {
          // Race: pick winner from fork stack
          const candidates = forkResults.pop() ?? [payload];
          if (candidates.length > 1) {
            const raceResult = await frameRace(
              candidates.map((candidate) => async () => candidate)
            );
            payload = raceResult.result;
          } else {
            payload = candidates[0];
          }
          break;
        }

        case FOLD: {
          // Fold: merge all fork results
          const toFold = forkResults.pop() ?? [payload];
          if (handler) {
            const ctx: GnosisHandlerContext = {
              nodeId,
              labels: [nodeId],
              properties: {},
              edges: [],
              engineRef: this,
              ...(authContext && { authContext, executionAuth: authContext }),
            };
            payload = await handler(toFold, {}, ctx);
          } else {
            payload = toFold.length === 1 ? toFold[0] : toFold;
          }
          break;
        }

        case VENT: {
          // Vent: dissipate, feed output back as new crossings
          if (handler) {
            const ctx: GnosisHandlerContext = {
              nodeId,
              labels: [nodeId],
              properties: {},
              edges: [],
              engineRef: this,
              ...(authContext && { authContext, executionAuth: authContext }),
            };
            const ventOutput = await handler(payload, {}, ctx);
            // Self-modification: if vent produces operation names, splice them back
            if (Array.isArray(ventOutput)) {
              const newCrossings = ventOutput
                .map((v: any) =>
                  typeof v === 'string' ? edgeTypeToCrossing(v) : null
                )
                .filter(
                  (c: any): c is import('./packed-knot.js').CrossingOp =>
                    c !== null
                );
              if (newCrossings.length > 0) {
                result = gen.next(newCrossings);
                continue;
              }
            }
          }
          logs.push(`[knot:${index}] VENT dissipated`);
          break;
        }
      }

      result = gen.next();
    }

    const law =
      this.activeLawState ??
      createLawStateFromCertificate(
        certifyGnosisAst(
          {
            nodes: new Map<string, ASTNode>(),
            edges: [],
          },
          {
            subject: 'gnosis-runtime-knot',
            metadata: {
              runtime: 'GnosisEngine.executeFromKnot',
              crossingCount: knot.crossingCount,
            },
          }
        ).certificate
      );

    return {
      logs: logs.join('\n'),
      payload,
      law,
      void: this.executionBoundary.snapshot(),
    };
  }
}
