// @a0n/gnosis/src/runtime/entanglement.ts
/**
 * Entanglement manager.
 *
 * Maintains a map of entangled VoidBoundary pairs. When one boundary
 * is updated, the complement distribution change propagates bidirectionally
 * to the entangled partner.
 */

import type { VoidBoundary, Resonance } from '../void';
import {
  createResonance,
  updateVoidBoundary,
  complementDistribution,
  boundaryDimensions,
  teleportDeficit,
} from '../void';

export interface EntanglementPair {
  /** First boundary in the entangled pair */
  boundaryA: VoidBoundary;
  /** Second boundary in the entangled pair */
  boundaryB: VoidBoundary;
  /** Coupling strength */
  strength: number;
  /** Resonance link */
  resonance: Resonance;
}

export class EntanglementManager {
  private pairs: EntanglementPair[] = [];

  /**
   * Entangle two boundaries. Updates to either will propagate to the other.
   */
  entangle(
    boundaryA: VoidBoundary,
    boundaryB: VoidBoundary,
    strength: number = 0.1
  ): EntanglementPair {
    const resonance = createResonance(0, 1, strength, 'entangled boundaries');
    const pair: EntanglementPair = {
      boundaryA,
      boundaryB,
      strength,
      resonance,
    };
    this.pairs.push(pair);
    return pair;
  }

  /**
   * Propagate complement distribution updates bidirectionally
   * for all entangled pairs.
   */
  propagate(): void {
    for (const pair of this.pairs) {
      this.propagatePair(pair);
    }
  }

  /**
   * Propagate updates for a single entangled pair.
   * Each boundary's complement distribution influences the other.
   */
  private propagatePair(pair: EntanglementPair): void {
    const distA = complementDistribution(pair.boundaryA);
    const distB = complementDistribution(pair.boundaryB);

    const dimsA = boundaryDimensions(pair.boundaryA);
    const dimsB = boundaryDimensions(pair.boundaryB);

    // A -> B: pull B toward A's complement shape
    if (dimsA === dimsB) {
      for (let i = 0; i < dimsB; i++) {
        pair.boundaryB.counts[i] += distA[i] * pair.strength;
      }
    } else {
      const ratio = dimsA / dimsB;
      for (let i = 0; i < dimsB; i++) {
        const srcIdx = Math.min(Math.floor(i * ratio), dimsA - 1);
        pair.boundaryB.counts[i] += distA[srcIdx] * pair.strength;
      }
    }
    pair.boundaryB.totalEntries = pair.boundaryB.counts.reduce(
      (a, b) => a + b,
      0
    );

    // B -> A: pull A toward B's complement shape
    if (dimsA === dimsB) {
      for (let i = 0; i < dimsA; i++) {
        pair.boundaryA.counts[i] += distB[i] * pair.strength;
      }
    } else {
      const ratio = dimsB / dimsA;
      for (let i = 0; i < dimsA; i++) {
        const srcIdx = Math.min(Math.floor(i * ratio), dimsB - 1);
        pair.boundaryA.counts[i] += distB[srcIdx] * pair.strength;
      }
    }
    pair.boundaryA.totalEntries = pair.boundaryA.counts.reduce(
      (a, b) => a + b,
      0
    );
  }

  /**
   * Get the deficit for teleportation from an entangled boundary.
   */
  teleportDeficit(boundary: VoidBoundary): number {
    return teleportDeficit(boundary);
  }

  /** Get all entangled pairs */
  get entangledPairs(): readonly EntanglementPair[] {
    return this.pairs;
  }

  /** Clear all entanglements */
  clear(): void {
    this.pairs = [];
  }
}
