// @a0n/gnosis/src/runtime/packed-knot.ts
/**
 * Bitwise DNA -- Packed Knot Binary Wire Format
 *
 * A PackedKnot with N crossings encodes exactly N topology operations
 * in 2 bits each. The knot formalizes the data. Zero overhead.
 *
 * Bitwise DNA encoding (four base operations, two bits each):
 *   00 = VENT   (dissipate)  -- stop codon
 *   01 = FOLD   (merge)      -- hydrogen bond
 *   10 = RACE   (select)     -- transcription
 *   11 = FORK   (split)      -- replication
 *
 * Four crossings per byte. Self-modifying: vent output feeds back
 * as input crossings. The topology rewrites itself during execution.
 *
 * Lean 4 theorems (FoldedKnotBitPacking.lean, 20, zero sorry):
 *   pack_unpack, random_access, zero_overhead, compose_assoc,
 *   sequence_matters, self_zero_dist, dist_symmetric, ...
 */

import type { ASTEdge, GraphAST } from '../betty/compiler';
import {
  MAGIC,
  COMMON_HEADER_SIZE,
  COMMON_HEADER_LARGE_SIZE,
  FLAG_LARGE,
  FLAG_RLE,
  FLAG_HOMOGENEOUS_IDS,
  writeHeader,
  readHeader,
  crc16,
  crc32,
  writeVarint,
  readVarint,
  verifyPayload,
} from '../lexicon';

// ─── Crossing codes (2-bit) ──────────────────────────────────────────

export type CrossingOp = 0 | 1 | 2 | 3;

export const VENT: CrossingOp = 0b00;
export const FOLD: CrossingOp = 0b01;
export const RACE: CrossingOp = 0b10;
export const FORK: CrossingOp = 0b11;

const CROSSING_NAMES: readonly string[] = ['VENT', 'FOLD', 'RACE', 'FORK'];

/**
 * Handles the gnosis crossing Name workflow.
 */
export function crossingName(op: CrossingOp): string {
  return CROSSING_NAMES[op] ?? '';
}

// ─── PackedKnot structure ────────────────────────────────────────────

export interface PackedKnot {
  crossingCount: number;
  crossings: Uint8Array; // 4 crossings per byte, big-endian within byte
  nodeIds: string[]; // node ID lookup by crossing index
}

/** DNA Mapping (Biochemical Alias) */
const DNA_BASES = ['A', 'C', 'G', 'T'];
const BASE_TO_OP: Record<string, CrossingOp> = {
  A: 0,
  C: 1,
  G: 2,
  T: 3,
};

/** Convert a PackedKnot to its DNA (ACGT) sequence. */
export function knotToDna(knot: PackedKnot): string {
  let dna = '';
  for (let i = 0; i < knot.crossingCount; i++) {
    dna += DNA_BASES[getCrossing(knot, i)];
  }
  return dna;
}

/** Convert a DNA (ACGT) sequence to a PackedKnot. */
export function dnaToKnot(dna: string): PackedKnot {
  const ops: CrossingOp[] = [];
  for (const char of dna.toUpperCase()) {
    const op = BASE_TO_OP[char];
    if (op !== undefined) ops.push(op);
  }
  // Use a single shared ID for DNA sequences to trigger homogeneous optimization
  return {
    crossingCount: ops.length,
    crossings: packCrossings(ops),
    nodeIds: new Array(ops.length).fill('_'),
  };
}

// ─── Edge type → crossing code ───────────────────────────────────────

/**
 * Handles the gnosis edge Type To Crossing workflow.
 */
export function edgeTypeToCrossing(type: string): CrossingOp {
  switch (type.toUpperCase()) {
    case 'FORK':
    case 'EVOLVE':
    case 'SUPERPOSE':
      return FORK;
    case 'RACE':
      return RACE;
    case 'FOLD':
    case 'COLLAPSE':
      return FOLD;
    case 'VENT':
    case 'TUNNEL':
      return VENT;
    // Default non-primitive edge types to FOLD (merge semantics)
    case 'PROCESS':
    case 'OBSERVE':
    case 'LAMINAR':
    case 'METACOG':
    case 'SLIVER':
      return FOLD;
    default:
      return FOLD;
  }
}

// ─── Random access (proven: random_access theorem) ───────────────────

/**
 * Handles the gnosis get Crossing workflow.
 */
export function getCrossing(knot: PackedKnot, i: number): CrossingOp {
  return (((knot.crossings[i >> 2] ?? 0) >> (6 - 2 * (i & 3))) & 3) as CrossingOp;
}

// ─── In-place mutation (self-modification) ───────────────────────────

/**
 * Handles the gnosis set Crossing workflow.
 */
export function setCrossing(knot: PackedKnot, i: number, op: CrossingOp): void {
  const byteIndex = i >> 2;
  const bitShift = 6 - 2 * (i & 3);
  knot.crossings[byteIndex] =
    ((knot.crossings[byteIndex] ?? 0) & ~(3 << bitShift)) | (op << bitShift);
}

// ─── Pack bits into bytes ────────────────────────────────────────────

function packCrossings(ops: CrossingOp[]): Uint8Array {
  const byteCount = Math.ceil(ops.length / 4);
  const bytes = new Uint8Array(byteCount);
  for (let i = 0; i < ops.length; i++) {
    const byteIndex = i >> 2;
    const bitShift = 6 - 2 * (i & 3);
    bytes[byteIndex] = (bytes[byteIndex] ?? 0) | (ops[i]! << bitShift);
  }
  return bytes;
}

// ─── Toposort ────────────────────────────────────────────────────────

function toposortEdges(edges: ASTEdge[]): ASTEdge[] {
  if (edges.length <= 1) return edges;

  // Build adjacency: for each node, which edges depend on it
  const inDegree = new Map<string, number>();
  const allNodes = new Set<string>();

  for (const edge of edges) {
    for (const sid of edge.sourceIds) allNodes.add(sid);
    for (const tid of edge.targetIds) allNodes.add(tid);
  }

  for (const node of allNodes) {
    inDegree.set(node, 0);
  }

  for (const edge of edges) {
    for (const tid of edge.targetIds) {
      inDegree.set(tid, (inDegree.get(tid) ?? 0) + 1);
    }
  }

  // Find roots (nodes with no incoming edges)
  const queue: string[] = [];
  for (const [node, deg] of inDegree) {
    if (deg === 0) queue.push(node);
  }

  // Map edges by source for ordering
  const edgeBySource = new Map<string, ASTEdge[]>();
  for (const edge of edges) {
    for (const sid of edge.sourceIds) {
      const list = edgeBySource.get(sid);
      if (list) list.push(edge);
      else edgeBySource.set(sid, [edge]);
    }
  }

  const sorted: ASTEdge[] = [];
  const visited = new Set<ASTEdge>();

  while (queue.length > 0) {
    const node = queue.shift()!;
    const nodeEdges = edgeBySource.get(node);
    if (!nodeEdges) continue;

    for (const edge of nodeEdges) {
      if (visited.has(edge)) continue;
      visited.add(edge);
      sorted.push(edge);

      for (const tid of edge.targetIds) {
        const deg = (inDegree.get(tid) ?? 1) - 1;
        inDegree.set(tid, deg);
        if (deg === 0) queue.push(tid);
      }
    }
  }

  // Append any unvisited edges (cycles -- encoded as-is)
  for (const edge of edges) {
    if (!visited.has(edge)) sorted.push(edge);
  }

  return sorted;
}

// ─── Reidemeister optimizations ──────────────────────────────────────

function reidemeisterOptimize(edges: ASTEdge[]): ASTEdge[] {
  const result: ASTEdge[] = [];

  for (let i = 0; i < edges.length; i++) {
    const edge = edges[i];
    if (!edge) continue;

    // R1: eliminate self-loops (source and target overlap)
    const sourceSet = new Set(edge.sourceIds);
    if (edge.targetIds.some((tid: string) => sourceSet.has(tid))) {
      continue;
    }

    // R2: cancel adjacent FORK-FOLD pairs targeting same nodes
    const nextEdge = edges[i + 1];
    if (
      nextEdge &&
      edgeTypeToCrossing(edge.type) === FORK &&
      edgeTypeToCrossing(nextEdge.type) === FOLD
    ) {
      const forkTargets = new Set(edge.targetIds);
      const foldSources = new Set(nextEdge.sourceIds);
      if (
        forkTargets.size === foldSources.size &&
        [...forkTargets].every((t) => foldSources.has(t))
      ) {
        i++; // skip the FOLD too
        continue;
      }
    }

    result.push(edge);
  }

  return result;
}

// ─── Coil: GraphAST → PackedKnot (the encoder) ──────────────────────

/**
 * Handles the gnosis coil workflow.
 */
export function coil(ast: GraphAST): PackedKnot {
  if (ast.edges.length === 0) {
    return { crossingCount: 0, crossings: new Uint8Array(0), nodeIds: [] };
  }

  // Stage 1: toposort
  const sorted = toposortEdges(ast.edges);

  // Stage 2: Reidemeister optimize
  const optimized = reidemeisterOptimize(sorted);

  // Stage 3: encode to 2-bit crossings
  const ops: CrossingOp[] = [];
  const nodeIds: string[] = [];

  for (const edge of optimized) {
    const op = edgeTypeToCrossing(edge.type);
    ops.push(op);
    nodeIds.push(edge.sourceIds[0] ?? '');
  }

  return {
    crossingCount: ops.length,
    crossings: packCrossings(ops),
    nodeIds,
  };
}

// ─── Release: PackedKnot → Iterator (the decoder) ───────────────────

export interface CrossingYield {
  op: CrossingOp;
  nodeId: string;
  index: number;
}

/**
 * Handles the gnosis release workflow.
 */
export function* release(
  knot: PackedKnot
): Generator<CrossingYield, void, CrossingOp[] | undefined> {
  let cursor = 0;
  let spliced: CrossingOp[] | undefined;

  while (cursor < knot.crossingCount) {
    const op = getCrossing(knot, cursor);
    const nodeId = knot.nodeIds[cursor] ?? '';

    // Yield crossing, accept optional splice-back from vent
    spliced = yield { op, nodeId, index: cursor };

    // Self-modification: vent output feeds back as new crossings
    if (spliced && spliced.length > 0) {
      const extended = composeKnots(knot, {
        crossingCount: spliced.length,
        crossings: packCrossings(spliced),
        nodeIds: spliced.map(() => nodeId),
      });
      knot.crossingCount = extended.crossingCount;
      knot.crossings = extended.crossings;
      knot.nodeIds = extended.nodeIds;
    }

    cursor++;
  }
}

// ─── Compose: concatenate knots (proven: compose_assoc) ──────────────

/**
 * Handles the gnosis compose Knots workflow.
 */
export function composeKnots(a: PackedKnot, b: PackedKnot): PackedKnot {
  const totalCount = a.crossingCount + b.crossingCount;
  const ops: CrossingOp[] = [];

  for (let i = 0; i < a.crossingCount; i++) ops.push(getCrossing(a, i));
  for (let i = 0; i < b.crossingCount; i++) ops.push(getCrossing(b, i));

  return {
    crossingCount: totalCount,
    crossings: packCrossings(ops),
    nodeIds: [...a.nodeIds, ...b.nodeIds],
  };
}

// ─── Hamming distance (proven: self_zero_dist, dist_symmetric) ───────

/**
 * Handles the gnosis hamming Distance workflow.
 */
export function hammingDistance(a: PackedKnot, b: PackedKnot): number {
  const len = Math.min(a.crossingCount, b.crossingCount);
  let dist = 0;
  for (let i = 0; i < len; i++) {
    if (getCrossing(a, i) !== getCrossing(b, i)) dist++;
  }
  return dist;
}

// ─── Serialization ───────────────────────────────────────────────────


/**
 * Converts input into Bytes.
 */
export function toBytes(
  knot: PackedKnot,
  options: { magic?: readonly [number, number] } = {}
): Uint8Array {
  const { magic = MAGIC.GNOT as readonly [number, number] } = options;

  // Collect unique node IDs (Holographic Dictionary)
  const uniqueIds: string[] = [];
  const idToIndex = new Map<string, number>();
  
  for (const id of knot.nodeIds) {
    if (!idToIndex.has(id)) {
      idToIndex.set(id, uniqueIds.length);
      uniqueIds.push(id);
    }
  }

  const encoder = new TextEncoder();
  const encodedUniqueIds = uniqueIds.map((id) => encoder.encode(id));
  
  // Density Fix: If all IDs are same, set FLAG_HOMOGENEOUS_IDS and omit index map
  const isHomogeneous = uniqueIds.length === 1 && knot.nodeIds.length === knot.crossingCount;
  const indexSize = uniqueIds.length <= 1 ? 0 : uniqueIds.length <= 256 ? 1 : uniqueIds.length <= 65536 ? 2 : 4;

  // Bits per crossing is 0.25 bytes
  const crossingBytes = Math.ceil(knot.crossingCount / 4);
  const tableEntriesSize = encodedUniqueIds.reduce((sum, e) => sum + 2 + e.length, 0);
  const indicesSize = isHomogeneous ? 0 : knot.crossingCount * indexSize;

  // We need a temporary buffer to calculate exact size - over-allocate safely
  const tempBuf = new Uint8Array(crossingBytes + tableEntriesSize + indicesSize + 256);
  let offset = 0;
  
  // Try RLE crossing compression if count is significant (Surface-Aware RLE)
  let bestCrossings = knot.crossings.subarray(0, crossingBytes);
  let useRLE = false;
  
  if (knot.crossingCount > 16) {
    const rleBuf = new Uint8Array(knot.crossingCount * 2); // worst case
    let rleOffset = 0;
    let runCount = 0;
    let lastOp = getCrossing(knot, 0);
    
    for (let i = 0; i < knot.crossingCount; i++) {
      const op = getCrossing(knot, i);
      if (op === lastOp && runCount < 0x3fffffff) {
        runCount++;
      } else {
        rleBuf[rleOffset++] = lastOp & 0x03;
        rleOffset += writeVarint(rleBuf, rleOffset, runCount);
        lastOp = op;
        runCount = 1;
      }
    }
    rleBuf[rleOffset++] = lastOp & 0x03;
    rleOffset += writeVarint(rleBuf, rleOffset, runCount);
    
    if (rleOffset < crossingBytes) {
      bestCrossings = rleBuf.subarray(0, rleOffset);
      useRLE = true;
    }
  }

  // Write crossingCount
  offset += writeVarint(tempBuf, offset, knot.crossingCount);
  
  // Write crossings
  if (useRLE) {
    offset += writeVarint(tempBuf, offset, bestCrossings.length);
  }
  tempBuf.set(bestCrossings, offset);
  offset += bestCrossings.length;
  
  // Write uniqueIdCount
  offset += writeVarint(tempBuf, offset, uniqueIds.length);
  
  // Write uniqueIdTable
  for (const b of encodedUniqueIds) {
    offset += writeVarint(tempBuf, offset, b.length);
    tempBuf.set(b, offset);
    offset += b.length;
  }
  // Only write index map if not homogeneous
  if (!isHomogeneous && uniqueIds.length > 0) {
    for (let i = 0; i < knot.crossingCount; i++) {
      const idx = idToIndex.get(knot.nodeIds[i] ?? '') ?? 0;
      if (indexSize === 1) {
        tempBuf[offset++] = idx & 0xff;
      } else if (indexSize === 2) {
        const view = new DataView(
          tempBuf.buffer,
          tempBuf.byteOffset,
          tempBuf.byteLength
        );
        view.setUint16(offset, idx, false);
        offset += 2;
      } else {
        const view = new DataView(
          tempBuf.buffer,
          tempBuf.byteOffset,
          tempBuf.byteLength
        );
        view.setUint32(offset, idx, false);
        offset += 4;
      }
    }
  }

  const innerPayload = tempBuf.subarray(0, offset);
  const innerLen = innerPayload.length;

  // Wrap in Lexicon Common Header
  const isLarge = innerLen > 0xffff;
  const headerSize = isLarge ? COMMON_HEADER_LARGE_SIZE : COMMON_HEADER_SIZE;
  const crc = isLarge ? crc32(innerPayload) : crc16(innerPayload);

  const buffer = new Uint8Array(headerSize + innerLen);
  writeHeader(buffer, magic, innerLen, {
    version: 3, // Version 3 is Surface-Aware
    flags:
      (isLarge ? FLAG_LARGE : 0) |
      (useRLE ? FLAG_RLE : 0) |
      (isHomogeneous ? FLAG_HOMOGENEOUS_IDS : 0),
    crc,
  });
  buffer.set(innerPayload, headerSize);

  return buffer;
}

/**
 * Handles the gnosis from Bytes workflow.
 */
export function fromBytes(bytes: Uint8Array): PackedKnot {
  const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);

  // Case-switch between Lexicon and legacy handled by header check
  const header = readHeader(bytes);
  if (
    header &&
    ( (header.magic[0] === MAGIC.GNOT[0] && header.magic[1] === MAGIC.GNOT[1]) ||
      (header.magic[0] === MAGIC.DNA[0] && header.magic[1] === MAGIC.DNA[1]) ) &&
    header.payloadLength > 0
  ) {
    const headerSize = (header.flags & FLAG_LARGE) !== 0
      ? COMMON_HEADER_LARGE_SIZE
      : COMMON_HEADER_SIZE;
      
    const innerPayload = bytes.subarray(headerSize, headerSize + header.payloadLength);

    // CRC verification (CRC=0 means uncomputed)
    if (header.crc !== 0) {
      if (!verifyPayload(header, innerPayload)) {
         throw new Error('PackedKnot: CRC verification failed');
      }
    }

    return decodeKnotInner(innerPayload, header.version, header.flags);
  }

  throw new Error(`Invalid knot magic: 0x${view.getUint16(0, false).toString(16)}`);
}

/** Decode the inner payload of a Lexicon-wrapped packed knot. */
function decodeKnotInner(
  innerPayload: Uint8Array,
  version: number,
  flags: number
): PackedKnot {
  const view = new DataView(
    innerPayload.buffer,
    innerPayload.byteOffset,
    innerPayload.byteLength
  );

  let offset = 0;
  const { value: crossingCount, bytesRead: bcRead } = readVarint(innerPayload, offset);
  offset += bcRead;
  
  const isRLE = (version >= 3) && (flags & FLAG_RLE);
  
  let crossings: Uint8Array;
  if (isRLE) {
    const { value: rleLen, bytesRead: rlRead } = readVarint(innerPayload, offset);
    offset += rlRead;
    const rleData = innerPayload.subarray(offset, offset + rleLen);
    offset += rleLen;
    
    // Decompress RLE to standard bit-packed
    const bitPacked = new Uint8Array(Math.ceil(crossingCount / 4));
    let bitOffset = 0;
    let rlePos = 0;
    while (rlePos < rleData.length) {
      const op = (rleData[rlePos++] ?? 0) & 0x03;
      const { value: count, bytesRead: cRead } = readVarint(rleData, rlePos);
      rlePos += cRead;
      
      for (let i = 0; i < count; i++) {
        const byteIndex = bitOffset >> 2;
        const bitShift = 6 - 2 * (bitOffset & 3);
        bitPacked[byteIndex] = (bitPacked[byteIndex] ?? 0) | (op << bitShift);
        bitOffset++;
      }
    }
    crossings = bitPacked;
  } else {
    const crossingBytes = Math.ceil(crossingCount / 4);
    crossings = innerPayload.slice(offset, offset + crossingBytes);
    offset += crossingBytes;
  }
  
  const decoder = new TextDecoder();

    // Version 3+: Holographic String Table + Varints
    const { value: uniqueIdCount, bytesRead: ucRead } = readVarint(innerPayload, offset);
    offset += ucRead;
    
    const uniqueIds: string[] = [];
    for (let i = 0; i < uniqueIdCount; i++) {
       const { value: len, bytesRead: lRead } = readVarint(innerPayload, offset);
       offset += lRead;
       uniqueIds.push(decoder.decode(innerPayload.subarray(offset, offset + len)));
       offset += len;
    }
    
    const nodeIds: string[] = [];
    const isHomogeneous = (flags & FLAG_HOMOGENEOUS_IDS) !== 0;

    if (isHomogeneous && uniqueIdCount > 0) {
      const singleId = uniqueIds[0] ?? '';
      for (let i = 0; i < crossingCount; i++) {
        nodeIds.push(singleId);
      }
    } else {
      const indexSize = uniqueIds.length <= 256 ? 1 : uniqueIds.length <= 65536 ? 2 : 4;
      for (let i = 0; i < crossingCount; i++) {
        let idx: number;
        if (indexSize === 1) {
          idx = innerPayload[offset++] ?? 0;
        } else if (indexSize === 2) {
          idx = view.getUint16(offset, false);
          offset += 2;
        } else {
          idx = view.getUint32(offset, false);
          offset += 4;
        }
        nodeIds.push(uniqueIds[idx] ?? '');
      }
    }
    return { crossingCount, crossings, nodeIds };
}

// ─── GENETIC EVOLUTION PRIMITIVES ────────────────────────────────────
// Topology-level evolution. The knot formalizes the genome.
// Slice, crossover, mutate, fitness. Self-optimizing code.

/**
 * Slice a contiguous sub-knot from crossing `start` to `end` (exclusive).
 * The sub-knot is a valid program by construction (compose_assoc).
 */
export function sliceKnot(
  knot: PackedKnot,
  start: number,
  end: number
): PackedKnot {
  const s = Math.max(0, Math.min(start, knot.crossingCount));
  const e = Math.max(s, Math.min(end, knot.crossingCount));
  const count = e - s;
  const ops: CrossingOp[] = [];
  for (let i = s; i < e; i++) ops.push(getCrossing(knot, i));
  return {
    crossingCount: count,
    crossings: packCrossings(ops),
    nodeIds: knot.nodeIds.slice(s, e),
  };
}

/**
 * Single-point crossover at crossing index `point`.
 * Parent A contributes crossings [0, point).
 * Parent B contributes crossings [point, end).
 * The offspring is a valid program (compose_assoc).
 */
export function crossover(
  a: PackedKnot,
  b: PackedKnot,
  point: number
): PackedKnot {
  const left = sliceKnot(a, 0, point);
  const right = sliceKnot(b, point, b.crossingCount);
  return composeKnots(left, right);
}

/**
 * Two-point crossover: take [0,p1) from A, [p1,p2) from B, [p2,end) from A.
 * Produces more diverse offspring than single-point.
 */
export function crossover2(
  a: PackedKnot,
  b: PackedKnot,
  p1: number,
  p2: number
): PackedKnot {
  const head = sliceKnot(a, 0, p1);
  const mid = sliceKnot(b, p1, p2);
  const tail = sliceKnot(a, p2, a.crossingCount);
  return composeKnots(composeKnots(head, mid), tail);
}

/**
 * Mutate: flip `count` random crossings to random operations.
 * Each mutation is a single bit-pair flip (2.5ns per mutation).
 * Returns the mutated knot (in-place, same reference).
 */
export function mutate(
  knot: PackedKnot,
  count: number,
  rng: () => number = Math.random
): PackedKnot {
  for (let i = 0; i < count; i++) {
    const idx = Math.floor(rng() * knot.crossingCount);
    const op = Math.floor(rng() * 4) as CrossingOp;
    setCrossing(knot, idx, op);
  }
  return knot;
}

/**
 * Fitness function type. Takes a knot, returns a score.
 * Higher is better. The knot formalizes the candidate program.
 */
export type FitnessFunction = (knot: PackedKnot) => number;

/**
 * Tournament selection: pick `tournamentSize` random knots from
 * the population, return the one with highest fitness.
 */
export function tournamentSelect(
  population: PackedKnot[],
  fitness: FitnessFunction,
  tournamentSize: number,
  rng: () => number = Math.random
): PackedKnot {
  let best: PackedKnot | null = null;
  let bestScore = -Infinity;
  for (let i = 0; i < tournamentSize; i++) {
    const candidate = population[Math.floor(rng() * population.length)];
    if (!candidate) continue;
    const score = fitness(candidate);
    if (score > bestScore) {
      bestScore = score;
      best = candidate;
    }
  }
  return best!;
}

/**
 * Run one generation of genetic evolution on a knot population.
 * Returns the next generation (same size).
 *
 * The topology evolves itself. Each knot is a program.
 * Crossover = byte concatenation. Mutation = bit flip.
 * Selection = execute and measure. The genome formalizes the program.
 */
export function evolveGeneration(
  population: PackedKnot[],
  fitness: FitnessFunction,
  options: {
    mutationRate?: number;
    mutationCount?: number;
    crossoverRate?: number;
    tournamentSize?: number;
    eliteCount?: number;
    rng?: () => number;
  } = {}
): PackedKnot[] {
  const {
    mutationRate = 0.1,
    mutationCount = 1,
    crossoverRate = 0.7,
    tournamentSize = 3,
    eliteCount = 1,
    rng = Math.random,
  } = options;

  const scored = population
    .map((knot) => ({ knot, score: fitness(knot) }))
    .sort((a, b) => b.score - a.score);

  const next: PackedKnot[] = [];

  // Elitism: keep top performers unchanged
  for (let i = 0; i < eliteCount && i < scored.length; i++) {
    next.push(scored[i]!.knot);
  }

  // Fill rest with crossover + mutation
  while (next.length < population.length) {
    const parentA = tournamentSelect(population, fitness, tournamentSize, rng);
    const parentB = tournamentSelect(population, fitness, tournamentSize, rng);

    let offspring: PackedKnot;
    if (rng() < crossoverRate) {
      const point = Math.floor(
        rng() * Math.min(parentA.crossingCount, parentB.crossingCount)
      );
      offspring = crossover(parentA, parentB, point);
    } else {
      // Clone parent A
      offspring = sliceKnot(parentA, 0, parentA.crossingCount);
    }

    if (rng() < mutationRate) {
      mutate(offspring, mutationCount, rng);
    }

    next.push(offspring);
  }

  return next;
}

// ─── VFS WRITEBACK ───────────────────────────────────────────────────
// Self-optimizing code writes its evolved state back to disk.
// The knot rewrites its own source file. The program evolves itself.

/**
 * Write a packed knot to a VFS path as a .knot binary.
 * The evolved topology persists across restarts.
 * The program rewrites its own genome.
 */
export async function writebackToVFS(
  knot: PackedKnot,
  filePath: string
): Promise<void> {
  const { writeFile } = await import('node:fs/promises');
  await writeFile(filePath, toBytes(knot));
}

/**
 * Load a packed knot from a VFS path.
 * The program reads its own genome from disk.
 */
export async function loadFromVFS(filePath: string): Promise<PackedKnot> {
  const { readFile } = await import('node:fs/promises');
  const bytes = new Uint8Array(await readFile(filePath));
  return fromBytes(bytes);
}

/**
 * Evolve a knot population, write the fittest back to VFS.
 * Self-optimizing loop: load → evolve → writeback → repeat.
 * The code improves itself across restarts.
 */
export async function evolveAndWriteback(
  filePath: string,
  fitness: FitnessFunction,
  options: {
    populationSize?: number;
    generations?: number;
    mutationRate?: number;
    crossoverRate?: number;
  } = {}
): Promise<{ best: PackedKnot; bestScore: number; generations: number }> {
  const {
    populationSize = 20,
    generations = 100,
    mutationRate = 0.1,
    crossoverRate = 0.7,
  } = options;

  // Load the seed knot from VFS
  const seed = await loadFromVFS(filePath);

  // Generate initial population by mutating the seed
  let population: PackedKnot[] = [seed];
  for (let i = 1; i < populationSize; i++) {
    const clone = sliceKnot(seed, 0, seed.crossingCount);
    mutate(clone, Math.max(1, Math.floor(seed.crossingCount * 0.1)));
    population.push(clone);
  }

  // Evolve
  for (let gen = 0; gen < generations; gen++) {
    population = evolveGeneration(population, fitness, {
      mutationRate,
      crossoverRate,
    });
  }

  // Find the best
  let best = population[0]!;
  let bestScore = fitness(best);
  for (const knot of population) {
    const score = fitness(knot);
    if (score > bestScore) {
      bestScore = score;
      best = knot;
    }
  }

  // Write the fittest back to VFS -- the program evolves itself
  await writebackToVFS(best, filePath);

  return { best, bestScore, generations };
}
