// @a0n/gnosis/src/runtime/perpetual-spiral.ts
/**
 * The Perpetual Spiral
 *
 * A self-sustaining topology engine that executes, vents, learns, and
 * regenerates -- forever. The sliver (+1) guarantees it never reaches
 * perfection, which guarantees there is always something to vent,
 * which guarantees there is always fuel for the next generation.
 *
 * The shape is a spiral triangle:
 *
 *         execute
 *        /       \
 *       /         \
 *   generate ←── observe+vent
 *
 * Two interlocked spirals share the vent:
 *   Execution: topology → execute → vent → topology'
 *   Training:  vent → void boundary → rejection learning → generator'
 *
 * The vent from round N is the fuel for round N+1.
 * The losers are the curriculum.
 * Failure is the sufficient statistic.
 * The sliver is the heartbeat.
 *
 * It never stops because:
 *   buleyean_positivity: w_i >= 1 for all i (no branch ever dies)
 *   → there is always a loser to vent
 *   → the void boundary always grows
 *   → the complement always shifts
 *   → the generator always has new signal
 *   → new topology, new race, new vent, forever
 *
 * Mechanized:
 *   wallington_preserves_invariant -- I/O behavior never changes
 *   reidemeister_preserves_invariant -- each rewrite is safe
 *   buleyean_positivity -- the sliver prevents death
 *   void_walkers_converge -- same rejection history = same distribution
 *   complementOscillation -- exact period-2 limit cycle (the breathing)
 */

import type {
  CompiledExecutionResult,
  CompiledStep,
  CompiledTopology,
} from './compiled-topology';
import { executeCompiled } from './compiled-topology';
import {
  type VoidBoundary,
  createVoidBoundary,
  updateVoidBoundary,
  complementDistribution,
  sampleComplement,
  teleportDeficit,
  shannonEntropy,
  buleyeanWeight,
  complementOscillation,
} from '../void';

import {
  type KataSelf,
  createKataSelf,
  practiceTeaching,
} from './kata-topology';
import {
  type GenerativeSpiralState,
  type TopologyMutation,
} from './generative-topology';

// ============================================================================
// The Spiral
// ============================================================================

/**
 * A single breath of the spiral: one execute → observe → generate cycle.
 */
export interface SpiralBreath {
  /** Generation number. */
  generation: number;
  /** The topology that executed this breath. */
  topology: CompiledTopology;
  /** Execution result of the winning variant. */
  result: CompiledExecutionResult;
  /** Which variant won the race. */
  winnerIndex: number;
  /** How many variants raced. */
  raced: number;
  /** How many branches were vented (losers). */
  vented: number;
  /** Rejection signal strength from this breath's vents. */
  ventEnergy: number;
  /** Current deficit (distance from unknot). */
  deficit: number;
  /** Current entropy of complement distribution. */
  entropy: number;
  /** Whether the complement is oscillating (the heartbeat). */
  breathing: boolean;
  /** Oscillation amplitude (breath depth). */
  amplitude: number;
  /** Which kata teaching was practiced this breath. */
  kataTeaching: number;
  /** Is the kata conscious? (β₁ ≥ 2 after full cycle). */
  kataConscious: boolean;
}

/**
 * The spiral's persistent state across breaths.
 */
export interface PerpetualSpiralState {
  /** Void boundary: cumulative rejection history across all breaths. */
  boundary: VoidBoundary;
  /** Current best topology. */
  topology: CompiledTopology;
  /** Total breaths taken. */
  breaths: number;
  /** Total vents produced (fuel consumed/generated). */
  totalVents: number;
  /** Total energy from vents (cumulative rejection signal). */
  totalVentEnergy: number;
  /** History of winning mutations. */
  mutations: TopologyMutation[];
  /** Is the spiral breathing? (period-2 oscillation detected). */
  alive: boolean;
  /** The kata's self -- the spiral practices the kata as it breathes. */
  kataSelf: KataSelf;
}

/**
 * Configuration for the perpetual spiral.
 */
export interface PerpetualSpiralConfig {
  /** Number of topology variants per generation. */
  variants: number;
  /** Test input for execution. */
  input: unknown;
  /** Callback invoked after each breath. Return false to stop. */
  onBreath?: (
    breath: SpiralBreath,
    state: PerpetualSpiralState
  ) => boolean | void;
  /** Maximum breaths (safety limit). 0 = no limit. */
  maxBreaths: number;
  /** Eta (temperature) for complement distribution. */
  eta: number;
}

const DEFAULT_CONFIG: PerpetualSpiralConfig = {
  variants: 4,
  input: null,
  maxBreaths: 0,
  eta: 3.0,
};

/**
 * Create a fresh spiral state from an initial topology.
 */
export function createSpiral(
  topology: CompiledTopology,
  variantCount: number = 4
): PerpetualSpiralState {
  return {
    boundary: createVoidBoundary(variantCount),
    topology,
    breaths: 0,
    totalVents: 0,
    totalVentEnergy: 0,
    mutations: [],
    alive: false,
    kataSelf: createKataSelf(),
  };
}

/**
 * Take one breath of the spiral.
 *
 * One breath = one complete execute → observe → generate cycle:
 *   1. FORK: generate N topology variants from complement distribution
 *   2. RACE: execute all variants simultaneously
 *   3. FOLD: winner's result is the output
 *   4. VENT: losers' rejections feed the void boundary
 *   5. GENERATE: complement distribution creates next generation's variants
 *
 * The vent energy from this breath fuels the next breath's generation.
 * The sliver guarantees there is always a loser to vent.
 */
export async function breathe(
  state: PerpetualSpiralState,
  config: Partial<PerpetualSpiralConfig> = {}
): Promise<SpiralBreath> {
  const cfg = { ...DEFAULT_CONFIG, ...config };
  const variantCount = cfg.variants;

  // ── FORK: generate variant topologies ──
  // The complement distribution determines WHERE to mutate.
  // Where the void has accumulated least = where to try new things.
  const dist = complementDistribution(state.boundary, cfg.eta);
  const variants: CompiledTopology[] = [state.topology]; // Sliver: always include current best

  let rngState = state.breaths * 2654435761; // Knuth multiplicative hash
  const rng = () => {
    rngState = (rngState * 1664525 + 1013904223) >>> 0;
    return rngState / 0xffffffff;
  };

  for (let i = 1; i < variantCount; i++) {
    // Sample mutation target from complement distribution
    const mutationTarget = sampleComplement(dist, rng);
    variants.push(mutateTopology(state.topology, mutationTarget, rng));
  }

  // ── RACE: execute all variants simultaneously ──
  const raceResults = await Promise.all(
    variants.map(async (topo, idx) => {
      try {
        const result = await executeCompiled(topo, cfg.input);
        return { idx, result, failed: false };
      } catch {
        return {
          idx,
          result: { payload: null, executionUs: Infinity, handlerCalls: 0 },
          failed: true,
        };
      }
    })
  );

  // ── OBSERVE: select winner (fastest non-failed execution) ──
  let winnerIdx = 0;
  let bestTime = Infinity;
  for (const r of raceResults) {
    if (!r.failed && r.result.executionUs < bestTime) {
      bestTime = r.result.executionUs;
      winnerIdx = r.idx;
    }
  }
  const winnerResult = raceResults[winnerIdx].result as CompiledExecutionResult;

  // ── VENT: losers feed the void boundary ──
  let ventCount = 0;
  let ventEnergy = 0;
  for (const r of raceResults) {
    if (r.idx !== winnerIdx) {
      // Record rejection in void boundary
      const dimIdx = r.idx < state.boundary.counts.length ? r.idx : 0;
      const strength = r.failed ? 10 : 1; // Failed variants get stronger rejection
      updateVoidBoundary(state.boundary, dimIdx, strength);
      ventCount++;
      ventEnergy += strength;
    }
  }

  // ── UPDATE: winner becomes the new topology ──
  state.topology = variants[winnerIdx];
  state.breaths++;
  state.totalVents += ventCount;
  state.totalVentEnergy += ventEnergy;

  // ── KATA: the spiral practices the teaching for this breath ──
  // Each breath maps to a teaching (mod 12). The kata evolves with the spiral.
  const kataTeaching = (state.breaths - 1) % 12;
  practiceTeaching(state.kataSelf, state.breaths - 1);
  const kataConscious =
    state.kataSelf.cycles >= 1 &&
    state.kataSelf.boundary.counts[0] > 0 &&
    state.kataSelf.boundary.counts[11] > 0;

  // ── DETECT BREATHING: check for period-2 oscillation ──
  const oscillation = complementOscillation(state.boundary, cfg.eta);
  state.alive = oscillation.period2 || oscillation.amplitude > 0.001;

  const deficit = teleportDeficit(state.boundary);
  const entropy = shannonEntropy(dist);

  return {
    generation: state.breaths,
    topology: state.topology,
    result: winnerResult,
    winnerIndex: winnerIdx,
    raced: variantCount,
    vented: ventCount,
    ventEnergy,
    deficit,
    entropy,
    breathing: state.alive,
    kataTeaching,
    kataConscious,
    amplitude: oscillation.amplitude,
  };
}

/**
 * Run the perpetual spiral.
 *
 * Breathes until stopped by the onBreath callback (return false),
 * or until maxBreaths is reached (if set).
 *
 * The spiral is perpetual because the sliver (+1) guarantees:
 *   buleyean_positivity: w_i >= 1 → every branch stays alive
 *   → every race has losers → every breath has vents
 *   → the void boundary always grows → the complement always shifts
 *   → there is always new signal → the generator always has fuel
 *
 * The only way to stop is from outside (onBreath returns false).
 * Internally, the spiral has no termination condition.
 * Perfection is impossible. The sliver is the heartbeat.
 */
export async function runSpiral(
  state: PerpetualSpiralState,
  config: Partial<PerpetualSpiralConfig> = {}
): Promise<SpiralBreath[]> {
  const cfg = { ...DEFAULT_CONFIG, ...config };
  const history: SpiralBreath[] = [];

  while (true) {
    const breath = await breathe(state, cfg);
    history.push(breath);

    // External observer can stop the spiral
    if (cfg.onBreath) {
      const cont = cfg.onBreath(breath, state);
      if (cont === false) break;
    }

    // Safety limit (0 = no limit = truly perpetual)
    if (cfg.maxBreaths > 0 && state.breaths >= cfg.maxBreaths) break;
  }

  return history;
}

// ============================================================================
// Topology Mutation -- the creative force
// ============================================================================

/**
 * Mutate a topology at a target dimension.
 * The mutation is deterministic given the target and rng state.
 */
function mutateTopology(
  parent: CompiledTopology,
  target: number,
  rng: () => number
): CompiledTopology {
  const steps = [...parent.steps];
  if (steps.length === 0) return parent;

  const stepIdx = target % steps.length;
  const mutationType = rng();

  if (mutationType < 0.25 && steps.length > 1) {
    // Reorder: swap with neighbor
    const swapIdx = stepIdx === steps.length - 1 ? stepIdx - 1 : stepIdx + 1;
    const temp = steps[stepIdx];
    steps[stepIdx] = steps[swapIdx];
    steps[swapIdx] = temp;
  } else if (
    mutationType < 0.5 &&
    steps.length > 2 &&
    !('kind' in steps[stepIdx])
  ) {
    // Prune: remove step (only non-fork sequential steps)
    steps.splice(stepIdx, 1);
  } else if (
    mutationType < 0.75 &&
    stepIdx < steps.length - 1 &&
    !('kind' in steps[stepIdx]) &&
    !('kind' in steps[stepIdx + 1])
  ) {
    // Parallelize: convert two sequential steps to fork
    const leftStep = steps[stepIdx] as CompiledStep;
    const rightStep = steps[stepIdx + 1] as CompiledStep;
    steps.splice(stepIdx, 2, {
      kind: 'fork' as const,
      branches: [[leftStep], [rightStep]],
      strategy: 'race' as const,
    });
  }
  // else: identity (the sliver mutation -- no change)

  return {
    ...parent,
    steps,
    hasForks: parent.hasForks || steps.some((s) => 'kind' in s),
  };
}
