// @a0n/gnosis/src/runtime/registry.ts
import type { GnosisExecutionAuthContext } from '../auth/core';
import type { GnosisCoreCache, GnosisCoreCacheSession } from './core-cache';
import type { LawState } from '../universe/index';

export interface GnosisHandlerContext {
  nodeId?: string;
  labels?: readonly string[];
  properties?: Readonly<Record<string, string>>;
  edges?: readonly unknown[];
  engineRef?: unknown;
  sharedState?: unknown;
  signal?: AbortSignal;
  authContext?: GnosisExecutionAuthContext;
  executionAuth?: GnosisExecutionAuthContext;
  coreCache?: GnosisCoreCache;
  cacheSession?: GnosisCoreCacheSession;
  law?: LawState;
}

export type GnosisHandler = (
  payload: any,
  props: Record<string, string>,
  context?: GnosisHandlerContext
) => Promise<any>;

export interface RegisterHandlerOptions {
  override?: boolean;
}

export class GnosisRegistry {
  private handlers = new Map<string, GnosisHandler>();

  public register(
    label: string,
    handler: GnosisHandler,
    options: RegisterHandlerOptions = {}
  ) {
    if (!options.override && this.handlers.has(label)) {
      return;
    }
    this.handlers.set(label, handler);
  }

  public hasHandler(label: string): boolean {
    return this.handlers.has(label);
  }

  public getHandler(label: string): GnosisHandler | undefined {
    return this.handlers.get(label);
  }

  public get(label: string): GnosisHandler | undefined {
    return this.getHandler(label);
  }
}
