// @a0n/gnosis/src/runtime/renderer-compat.ts
type TopologyCollapseStrategy = 'race' | 'fold';

interface TopologyBranch {
  id: string;
  vertices: number;
  drawCalls: number;
  estimatedCostMs: number;
  quality?: number;
  vent?: boolean;
}

interface TopologyFrameInput {
  branches: TopologyBranch[];
  strategy?: TopologyCollapseStrategy;
  budgetMs?: number;
}

interface TopologyFrameResult {
  collapsedBy: TopologyCollapseStrategy;
  winnerId: string | null;
  survivingBranches: TopologyBranch[];
  ventedBranchIds: string[];
  totalVertices: number;
  totalDrawCalls: number;
  totalCostMs: number;
  budgetMs: number;
  overBudget: boolean;
}

type ExecuteTopologyFrame = (
  input: TopologyFrameInput
) => Promise<TopologyFrameResult>;

interface GnosisSceneBranch {
  id?: unknown;
  vertices?: unknown;
  drawCalls?: unknown;
  estimatedCostMs?: unknown;
  quality?: unknown;
  vent?: unknown;
}

interface GnosisScenePayload {
  branches: GnosisSceneBranch[];
  budgetMs?: unknown;
  collapse?: unknown;
}

export interface GnosisTopologyRenderOutput {
  renderer: 'aeon-3d-compat';
  source: 'aeon-3d' | 'gnosis-fallback';
  frame: TopologyFrameResult;
  summary: {
    forked: number;
    vented: number;
    collapsedBy: TopologyCollapseStrategy;
  };
}

type ResolvedRendererEngine = {
  source: 'aeon-3d' | 'gnosis-fallback';
  executeTopologyFrame: ExecuteTopologyFrame;
};

const AEON_3D_MODULE_NAME = '@a0n/aeon-3d';

let cachedRendererEngine: ResolvedRendererEngine | null = null;

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null;
}

function parseNumber(value: unknown, fallback: number): number {
  if (typeof value === 'number' && Number.isFinite(value)) {
    return value;
  }
  if (typeof value === 'string') {
    const parsed = Number.parseFloat(value);
    if (Number.isFinite(parsed)) {
      return parsed;
    }
  }
  return fallback;
}

function parseBoolean(value: unknown, fallback = false): boolean {
  if (typeof value === 'boolean') {
    return value;
  }
  if (typeof value === 'string') {
    const lowered = value.toLowerCase();
    if (lowered === 'true') return true;
    if (lowered === 'false') return false;
  }
  return fallback;
}

function parseStrategy(value: unknown): TopologyCollapseStrategy | null {
  if (typeof value !== 'string') {
    return null;
  }
  const normalized = value.toLowerCase();
  if (normalized === 'race' || normalized === 'fold') {
    return normalized;
  }
  return null;
}

function normalizeBranch(
  branch: GnosisSceneBranch,
  index: number
): TopologyBranch {
  const fallbackId = `branch-${index + 1}`;
  const id =
    typeof branch.id === 'string' && branch.id.trim().length > 0
      ? branch.id.trim()
      : fallbackId;

  return {
    id,
    vertices: Math.max(0, Math.round(parseNumber(branch.vertices, 10_000))),
    drawCalls: Math.max(0, Math.round(parseNumber(branch.drawCalls, 12))),
    estimatedCostMs: Math.max(0, parseNumber(branch.estimatedCostMs, 2.0)),
    quality: Math.max(0.1, parseNumber(branch.quality, 1)),
    vent: parseBoolean(branch.vent, false),
  };
}

function parseScenePayload(payload: unknown): GnosisScenePayload | null {
  if (!isRecord(payload)) {
    return null;
  }
  if (!isRecord(payload.scene)) {
    return null;
  }
  const maybeBranches = payload.scene['branches'];
  if (!Array.isArray(maybeBranches)) {
    return null;
  }
  return {
    branches: maybeBranches,
    budgetMs: payload.scene['budgetMs'],
    collapse: payload.scene['collapse'],
  };
}

function toHtml(payload: unknown): string {
  const theme =
    isRecord(payload) && typeof payload.theme === 'string'
      ? payload.theme
      : 'light';
  return `<html><body data-theme="${theme}"><h1>Aeon Flux</h1></body></html>`;
}

function ventBranches(
  branches: TopologyBranch[],
  budgetMs: number
): { active: TopologyBranch[]; ventedBranchIds: string[] } {
  const ventedBranchIds: string[] = [];
  const explicitlyActive = branches.filter((branch) => {
    if (branch.vent) {
      ventedBranchIds.push(branch.id);
      return false;
    }
    return true;
  });

  const sortedByCost = [...explicitlyActive].sort(
    (left, right) => left.estimatedCostMs - right.estimatedCostMs
  );

  const active: TopologyBranch[] = [];
  let runningCost = 0;
  for (const branch of sortedByCost) {
    const nextCost = runningCost + branch.estimatedCostMs;
    if (nextCost > budgetMs && active.length > 0) {
      ventedBranchIds.push(branch.id);
      continue;
    }
    active.push(branch);
    runningCost = nextCost;
  }

  return { active, ventedBranchIds };
}

const executeTopologyFrameFallback: ExecuteTopologyFrame = async (
  input: TopologyFrameInput
) => {
  const budgetMs = Math.max(0.1, parseNumber(input.budgetMs, 8));
  const strategy = input.strategy ?? 'fold';
  const branches = input.branches.map((branch, index) =>
    normalizeBranch(branch, index)
  );

  const { active, ventedBranchIds } = ventBranches(branches, budgetMs);
  const survivingBranches = active.length > 0 ? active : branches.slice(0, 1);

  let collapsedBy: TopologyCollapseStrategy = strategy;
  let winnerId: string | null = null;
  let collapsedBranches = survivingBranches;

  if (strategy === 'race') {
    const winner = [...survivingBranches].sort((left, right) => {
      const leftScore = left.estimatedCostMs / (left.quality ?? 1);
      const rightScore = right.estimatedCostMs / (right.quality ?? 1);
      return leftScore - rightScore;
    })[0];

    if (winner) {
      winnerId = winner.id;
      collapsedBranches = [winner];
    } else {
      collapsedBy = 'fold';
    }
  }

  const totals = collapsedBranches.reduce(
    (accumulator, branch) => {
      accumulator.vertices += branch.vertices;
      accumulator.drawCalls += branch.drawCalls;
      accumulator.cost += branch.estimatedCostMs;
      return accumulator;
    },
    { vertices: 0, drawCalls: 0, cost: 0 }
  );

  return {
    collapsedBy,
    winnerId,
    survivingBranches: collapsedBranches,
    ventedBranchIds: Array.from(new Set(ventedBranchIds)),
    totalVertices: totals.vertices,
    totalDrawCalls: totals.drawCalls,
    totalCostMs: Number(totals.cost.toFixed(3)),
    budgetMs,
    overBudget: totals.cost > budgetMs,
  };
};

async function resolveRendererEngine(): Promise<ResolvedRendererEngine> {
  if (cachedRendererEngine) {
    return cachedRendererEngine;
  }

  try {
    const moduleName: string = AEON_3D_MODULE_NAME;
    const loadedModule = await import(moduleName);
    const maybeExecute = isRecord(loadedModule)
      ? loadedModule['executeTopologyFrame']
      : null;
    if (typeof maybeExecute === 'function') {
      const executeTopologyFrame = maybeExecute as ExecuteTopologyFrame;
      cachedRendererEngine = {
        source: 'aeon-3d',
        executeTopologyFrame,
      };
      return cachedRendererEngine;
    }
  } catch {
    // no-op: fallback engine handles local/dev mode where aeon-3d isn't installed.
  }

  cachedRendererEngine = {
    source: 'gnosis-fallback',
    executeTopologyFrame: executeTopologyFrameFallback,
  };
  return cachedRendererEngine;
}

/**
 * Handles the gnosis render With Topology Compat workflow.
 */
export async function renderWithTopologyCompat(
  payload: unknown,
  props: Record<string, string>
): Promise<string | GnosisTopologyRenderOutput> {
  const scene = parseScenePayload(payload);
  const requestedType =
    typeof props.type === 'string' ? props.type.toLowerCase() : null;
  const effectiveType = requestedType ?? (scene ? '3d' : 'html');

  if (effectiveType !== '3d' && effectiveType !== 'topology') {
    return toHtml(payload);
  }

  const branches = (scene?.branches ?? []).map((branch, index) =>
    normalizeBranch(branch, index)
  );
  const budgetFromScene = scene ? parseNumber(scene.budgetMs, 8) : 8;
  const budgetMs = parseNumber(props.budgetMs, budgetFromScene);
  const collapseStrategy =
    parseStrategy(props.collapse) ??
    parseStrategy(props.strategy) ??
    parseStrategy(scene?.collapse) ??
    'fold';

  const { source, executeTopologyFrame } = await resolveRendererEngine();
  const frame = await executeTopologyFrame({
    branches,
    strategy: collapseStrategy,
    budgetMs,
  });

  return {
    renderer: 'aeon-3d-compat',
    source,
    frame,
    summary: {
      forked: branches.length,
      vented: frame.ventedBranchIds.length,
      collapsedBy: frame.collapsedBy,
    },
  };
}
