// @a0n/gnosis/src/runtime/structured-concurrency.ts
export type StructuredConcurrencyFailureMode = 'cancel' | 'vent' | 'shield';

export type StructuredBranchStatus =
  | 'success'
  | 'error'
  | 'timeout'
  | 'cancelled'
  | 'vented';

export interface StructuredConcurrencyPolicy {
  readonly failure: StructuredConcurrencyFailureMode;
  readonly timeoutMs: number | null;
  readonly deadlineMs: number | null;
}

export interface ConcurrentBranchOutcome {
  readonly path: string;
  readonly sid: number;
  readonly status: StructuredBranchStatus;
  readonly durationMs: number;
  readonly value?: unknown;
  readonly reason?: string;
}

export interface ConcurrentBranchHandle {
  readonly path: string;
  readonly sid: number;
  readonly controller: AbortController;
  promise: Promise<ConcurrentBranchOutcome>;
  settled: boolean;
  outcome?: ConcurrentBranchOutcome;
}

export interface StructuredRaceResolution {
  readonly winner: ConcurrentBranchOutcome | null;
  readonly outcomes: readonly ConcurrentBranchOutcome[];
  readonly cancelledByFailure: ConcurrentBranchOutcome | null;
}

export interface StructuredFoldResolution {
  readonly outcomes: readonly ConcurrentBranchOutcome[];
  readonly cancelledByFailure: ConcurrentBranchOutcome | null;
}

function trackPendingBranches(
  branches: readonly ConcurrentBranchHandle[]
): Map<string, Promise<ConcurrentBranchOutcome>> {
  return new Map(
    branches.map((branch) => [branch.path, branch.promise] as const)
  );
}

/**
 * Handles the gnosis collect Concurrent Outcomes workflow.
 */
export async function collectConcurrentOutcomes(
  branches: readonly ConcurrentBranchHandle[]
): Promise<readonly ConcurrentBranchOutcome[]> {
  return Promise.all(branches.map((branch) => branch.promise));
}

/**
 * Handles the gnosis cancel Concurrent Branches workflow.
 */
export function cancelConcurrentBranches(
  branches: readonly ConcurrentBranchHandle[],
  reason: string,
  survivorPath?: string
): void {
  for (const branch of branches) {
    if (
      branch.path === survivorPath ||
      branch.settled ||
      branch.controller.signal.aborted
    ) {
      continue;
    }

    branch.controller.abort({
      kind: 'cancelled',
      reason,
    });
  }
}

/**
 * Handles the gnosis map Outcomes For Policy workflow.
 */
export function mapOutcomesForPolicy(
  outcomes: readonly ConcurrentBranchOutcome[],
  policy: StructuredConcurrencyPolicy
): readonly ConcurrentBranchOutcome[] {
  if (policy.failure !== 'vent') {
    return outcomes;
  }

  return outcomes.map((outcome) => {
    if (outcome.status === 'error' || outcome.status === 'timeout') {
      return {
        ...outcome,
        status: 'vented' as const,
      };
    }

    return outcome;
  });
}

/**
 * Resolves the Structured Race.
 */
export async function resolveStructuredRace(
  branches: readonly ConcurrentBranchHandle[],
  policy: StructuredConcurrencyPolicy
): Promise<StructuredRaceResolution> {
  const pending = trackPendingBranches(branches);
  let winner: ConcurrentBranchOutcome | null = null;
  let cancelledByFailure: ConcurrentBranchOutcome | null = null;

  while (pending.size > 0) {
    const settled = await Promise.race(
      Array.from(pending.entries()).map(([path, promise]) =>
        promise.then((outcome) => ({ path, outcome }))
      )
    );
    pending.delete(settled.path);

    if (settled.outcome.status === 'success') {
      winner = settled.outcome;
      cancelConcurrentBranches(
        branches,
        `cancelled after RACE winner '${settled.outcome.path}'.`,
        settled.outcome.path
      );
      break;
    }

    if (
      (settled.outcome.status === 'error' ||
        settled.outcome.status === 'timeout') &&
      policy.failure === 'cancel'
    ) {
      cancelledByFailure = settled.outcome;
      cancelConcurrentBranches(
        branches,
        `cancelled after ${settled.outcome.status} in '${settled.outcome.path}'.`,
        settled.outcome.path
      );
      break;
    }
  }

  return {
    winner,
    outcomes: mapOutcomesForPolicy(
      await collectConcurrentOutcomes(branches),
      policy
    ),
    cancelledByFailure,
  };
}

/**
 * Resolves the Structured Fold.
 */
export async function resolveStructuredFold(
  branches: readonly ConcurrentBranchHandle[],
  policy: StructuredConcurrencyPolicy
): Promise<StructuredFoldResolution> {
  const pending = trackPendingBranches(branches);
  let cancelledByFailure: ConcurrentBranchOutcome | null = null;

  while (pending.size > 0) {
    const settled = await Promise.race(
      Array.from(pending.entries()).map(([path, promise]) =>
        promise.then((outcome) => ({ path, outcome }))
      )
    );
    pending.delete(settled.path);

    if (
      (settled.outcome.status === 'error' ||
        settled.outcome.status === 'timeout') &&
      policy.failure === 'cancel'
    ) {
      cancelledByFailure = settled.outcome;
      cancelConcurrentBranches(
        branches,
        `cancelled after ${settled.outcome.status} in '${settled.outcome.path}'.`,
        settled.outcome.path
      );
      break;
    }
  }

  return {
    outcomes: mapOutcomesForPolicy(
      await collectConcurrentOutcomes(branches),
      policy
    ),
    cancelledByFailure,
  };
}
