// @a0n/gnosis/src/runtime/subprocess-invoker.ts
import { existsSync } from 'node:fs';
import { spawn, type ChildProcess } from 'node:child_process';
import * as path from 'node:path';
import { fileURLToPath } from 'node:url';

/**
 * Request to invoke a function in a target language via a subprocess harness.
 */
export interface InvokeCompatibilityBody {
  encoding: 'base64';
  value: string;
}

export interface InvokeCompatibilityHttpRequest {
  method: string;
  path: string;
  query?: string;
  headers: Record<string, string>;
  body?: InvokeCompatibilityBody | null;
}

export interface NodeRouterCompatibilityRequest {
  family: 'node-router';
  framework: string;
  request: InvokeCompatibilityHttpRequest;
}

export interface GoRouterCompatibilityRequest {
  family: 'go-router';
  framework: string;
  request: InvokeCompatibilityHttpRequest;
}

export interface PythonServiceCompatibilityRequest {
  family: 'python-service';
  framework: string;
  request: InvokeCompatibilityHttpRequest;
}

export interface RubyServiceCompatibilityRequest {
  family: 'ruby-service';
  framework: string;
  request: InvokeCompatibilityHttpRequest;
}

export interface JavaServiceCompatibilityRequest {
  family: 'java-service';
  framework: string;
  request: InvokeCompatibilityHttpRequest;
}

export interface ScriptServiceCompatibilityRequest {
  family: 'script-service';
  framework: 'standard';
  request: InvokeCompatibilityHttpRequest;
}

export interface NodeRouterCompatibilityResponse {
  matched: boolean;
  response?: {
    status: number;
    headers: Record<string, string>;
    body?: InvokeCompatibilityBody | null;
  };
}

export interface GoRouterCompatibilityResponse {
  matched: boolean;
  response?: {
    status: number;
    headers: Record<string, string>;
    body?: InvokeCompatibilityBody | null;
  };
}

export interface PythonServiceCompatibilityResponse {
  matched: boolean;
  response?: {
    status: number;
    headers: Record<string, string>;
    body?: InvokeCompatibilityBody | null;
  };
}

export interface RubyServiceCompatibilityResponse {
  matched: boolean;
  response?: {
    status: number;
    headers: Record<string, string>;
    body?: InvokeCompatibilityBody | null;
  };
}

export interface JavaServiceCompatibilityResponse {
  matched: boolean;
  response?: {
    status: number;
    headers: Record<string, string>;
    body?: InvokeCompatibilityBody | null;
  };
}

export interface ScriptServiceCompatibilityResponse {
  matched: boolean;
  response?: {
    status: number;
    headers: Record<string, string>;
    body?: InvokeCompatibilityBody | null;
  };
}

export type InvokeCompatibilityRequest =
  | NodeRouterCompatibilityRequest
  | GoRouterCompatibilityRequest
  | PythonServiceCompatibilityRequest
  | RubyServiceCompatibilityRequest
  | JavaServiceCompatibilityRequest
  | ScriptServiceCompatibilityRequest;

export interface InvokeRequest {
  action: 'execute' | 'ping';
  language: string;
  filePath: string;
  functionName: string;
  args: unknown[];
  sourceRange?: { startByte: number; endByte: number };
  compatibility?: InvokeCompatibilityRequest;
}

/**
 * Response from a subprocess harness execution.
 */
export interface InvokeResponse {
  status: 'ok' | 'error';
  value: unknown;
  stdout: string;
  stderr: string;
  /** Mapped error with language-aware context (populated on error). */
  mappedError?: MappedError;
}

/**
 * Language-aware error mapping -- traces errors back to source locations.
 */
export interface MappedError {
  language: string;
  filePath: string;
  functionName: string;
  message: string;
  /** Extracted line number from traceback/backtrace if available. */
  line?: number;
  /** Extracted column from traceback if available. */
  column?: number;
  /** Original traceback/backtrace text. */
  traceback: string;
  /** Source range that was being executed. */
  sourceRange?: { startByte: number; endByte: number };
}

/**
 * Default language-to-command mapping.
 * Each entry is [command, ...prefixArgs].
 */
const DEFAULT_LANGUAGE_COMMANDS: Record<string, string[]> = {
  // Phase 1: Hand-written extractors
  python: ['python3'],
  javascript: ['node'],
  typescript: ['node', '--import', 'tsx'],
  go: ['go', 'run'],
  ruby: ['ruby'],
  java: ['node'],
  rust: ['sh'],
  // Phase 2: Declarative extractors
  c: ['sh'],
  cpp: ['sh'],
  c_sharp: ['sh'],
  kotlin: ['sh'],
  scala: ['sh'],
  swift: ['sh'],
  haskell: ['sh'],
  ocaml: ['sh'],
  lua: ['lua'],
  php: ['php'],
  elixir: ['elixir'],
  zig: ['sh'],
  bash: ['sh'],
  shell: ['sh'],
};

/**
 * Resolve the harness script path for a given language.
 */
function resolveHarnessPath(language: string): string {
  const harnessMap: Record<string, string> = {
    // Phase 1: Dedicated harnesses
    python: 'python_harness.py',
    javascript: 'node_harness.mjs',
    typescript: 'node_harness.mjs',
    go: 'go_harness.go',
    ruby: 'ruby_harness.rb',
    lua: 'lua_harness.lua',
    php: 'php_harness.php',
    elixir: 'elixir_harness.exs',
    // Phase 2: Language-specific shell harnesses
    kotlin: 'kotlin_harness.sh',
    scala: 'scala_harness.sh',
    swift: 'swift_harness.sh',
    haskell: 'haskell_harness.sh',
    ocaml: 'ocaml_harness.sh',
    zig: 'zig_harness.sh',
    c_sharp: 'csharp_harness.sh',
    bash: 'bash_harness.sh',
    // Fallback to generic
    c: 'generic_harness.sh',
    cpp: 'generic_harness.sh',
    rust: 'generic_harness.sh',
    java: 'java_harness.mjs',
    shell: 'generic_harness.sh',
    'script-service': 'script_service_harness.py',
  };

  const harnessFile = harnessMap[language] ?? 'generic_harness.sh';
  const importMetaHarnessDir = path.resolve(
    path.dirname(fileURLToPath(import.meta.url)),
    '..',
    '..',
    'harnesses'
  );
  const candidatePaths = [
    path.join(importMetaHarnessDir, harnessFile),
    path.resolve(process.cwd(), 'harnesses', harnessFile),
    path.resolve(process.cwd(), '..', 'gnosis', 'harnesses', harnessFile),
    path.resolve(
      process.cwd(),
      'open-source',
      'gnosis',
      'harnesses',
      harnessFile
    ),
  ];

  return (
    candidatePaths.find((candidatePath) => existsSync(candidatePath)) ??
    candidatePaths[0] ??
    path.join(importMetaHarnessDir, harnessFile)
  );
}

/**
 * Get the command to launch for a given language, respecting env overrides.
 * Override pattern: GNODE_PYTHON_CMD, GNODE_GO_CMD, etc.
 */
function getLanguageCommand(language: string): string[] {
  const envKey = `GNODE_${language.toUpperCase()}_CMD`;
  const envCmd = process.env[envKey];
  if (envCmd) {
    return envCmd.split(/\s+/);
  }
  return DEFAULT_LANGUAGE_COMMANDS[language] ?? ['sh'];
}

interface InvocationHarnessRuntime {
  readonly harnessPath: string;
  readonly commandParts: string[];
}

function resolveInvocationHarness(
  request: InvokeRequest
): InvocationHarnessRuntime {
  if (request.compatibility?.family === 'script-service') {
    return {
      harnessPath: resolveHarnessPath('script-service'),
      commandParts: ['python3'],
    };
  }

  return {
    harnessPath: resolveHarnessPath(request.language),
    commandParts: getLanguageCommand(request.language),
  };
}

// ─── Warm Process Pool ───────────────────────────────────────────────────────
// Keeps one subprocess per language alive for repeated calls.
// Interpreted languages (Python, Ruby, Lua, Node) benefit most from pooling
// since interpreter startup is the dominant cost.

interface WarmPoolEntry {
  process: ChildProcess;
  busy: boolean;
  language: string;
  lastUsed: number;
}

const warmPool = new Map<string, WarmPoolEntry>();

/** Pool cleanup interval. */
let poolCleanupTimer: ReturnType<typeof setInterval> | null = null;

function killWarmPool(): void {
  for (const [, entry] of warmPool) {
    entry.process.kill('SIGTERM');
  }
  warmPool.clear();
  if (poolCleanupTimer) {
    clearInterval(poolCleanupTimer);
    poolCleanupTimer = null;
  }
}

// Clean up on exit.
process.on('exit', killWarmPool);

/**
 * Get the subprocess timeout in milliseconds.
 */
function getTimeoutMs(request?: InvokeRequest): number {
  const envVal = process.env.GNODE_SUBPROCESS_TIMEOUT_MS;
  if (envVal) {
    const parsed = Number.parseInt(envVal, 10);
    if (Number.isFinite(parsed) && parsed > 0) {
      return parsed;
    }
  }
  if (request?.compatibility?.family === 'go-router') {
    return 60_000;
  }
  if (request?.compatibility?.family === 'java-service') {
    return 60_000;
  }
  return 30_000; // 30s default
}

// ─── Language-Aware Error Mapping ────────────────────────────────────────────

/**
 * Extract line/column from a language-specific error traceback.
 */
function mapErrorToSource(
  language: string,
  filePath: string,
  functionName: string,
  errorValue: unknown,
  stderr: string,
  sourceRange?: { startByte: number; endByte: number }
): MappedError {
  const errorStr =
    typeof errorValue === 'string' ? errorValue : String(errorValue);
  const fullText = `${errorStr}\n${stderr}`;

  let line: number | undefined;
  let column: number | undefined;

  const escapedFilePath = filePath.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');

  // Python: File "path.py", line 42, in func_name
  const pyMatch =
    fullText.match(new RegExp(`File "${escapedFilePath}", line (\\d+)`)) ??
    fullText.match(/File "([^"]+)", line (\d+)/);
  if (pyMatch) {
    const pyLine = pyMatch[2] ?? pyMatch[1];
    if (pyLine) {
      line = Number.parseInt(pyLine, 10);
    }
  }

  // Go: path.go:42:5: error message
  const goMatch = fullText.match(/\.go:(\d+):(\d+)/);
  if (goMatch) {
    const goLine = goMatch[1];
    const goColumn = goMatch[2];
    if (goLine) line = Number.parseInt(goLine, 10);
    if (goColumn) column = Number.parseInt(goColumn, 10);
  }

  // Rust: --> src/main.rs:42:5
  const rsMatch = fullText.match(/--> [^:]+:(\d+):(\d+)/);
  if (rsMatch) {
    const rsLine = rsMatch[1];
    const rsColumn = rsMatch[2];
    if (rsLine) line = Number.parseInt(rsLine, 10);
    if (rsColumn) column = Number.parseInt(rsColumn, 10);
  }

  // Ruby: path.rb:42:in `method': message (ErrorClass)
  const rbMatch = fullText.match(/\.rb:(\d+):in/);
  if (rbMatch) {
    const rbLine = rbMatch[1];
    if (rbLine) line = Number.parseInt(rbLine, 10);
  }

  // Java: at Class.method(File.java:42)
  const javaMatch = fullText.match(/\.java:(\d+)\)/);
  if (javaMatch) {
    const javaLine = javaMatch[1];
    if (javaLine) line = Number.parseInt(javaLine, 10);
  }

  // JavaScript/TypeScript: at function (path.js:42:5)
  const jsMatch = fullText.match(/\.(js|ts|mjs|cjs):(\d+):(\d+)/);
  if (jsMatch) {
    const jsLine = jsMatch[2];
    const jsColumn = jsMatch[3];
    if (jsLine) line = Number.parseInt(jsLine, 10);
    if (jsColumn) column = Number.parseInt(jsColumn, 10);
  }

  // C/C++: path.c:42: error
  const cMatch = fullText.match(/\.(c|cpp|cc|h):(\d+)/);
  if (cMatch && !line) {
    const cLine = cMatch[2];
    if (cLine) line = Number.parseInt(cLine, 10);
  }

  // Swift: path.swift:42:5: error
  const swiftMatch = fullText.match(/\.swift:(\d+):(\d+)/);
  if (swiftMatch && !line) {
    const swiftLine = swiftMatch[1];
    const swiftColumn = swiftMatch[2];
    if (swiftLine) line = Number.parseInt(swiftLine, 10);
    if (swiftColumn) column = Number.parseInt(swiftColumn, 10);
  }

  // Extract first meaningful error message line.
  const message =
    errorStr.split('\n').find((l) => l.trim().length > 0) ?? errorStr;

  const mappedError: MappedError = {
    language,
    filePath,
    functionName,
    message,
    traceback: errorStr,
  };
  if (line !== undefined) mappedError.line = line;
  if (column !== undefined) mappedError.column = column;
  if (sourceRange !== undefined) mappedError.sourceRange = sourceRange;
  return mappedError;
}

// ─── Core Invocation ─────────────────────────────────────────────────────────

/**
 * Invoke a function in a target language via its subprocess harness.
 *
 * Protocol: send JSON request on stdin, read JSON response from stdout.
 * stderr is captured separately for debugging.
 */
export async function invokeSubprocess(
  request: InvokeRequest
): Promise<InvokeResponse> {
  const timeoutMs = getTimeoutMs(request);
  const { harnessPath, commandParts } = resolveInvocationHarness(request);

  const [cmd, ...commandTail] = commandParts;
  if (!cmd) {
    throw new Error(`No launch command configured for language ${request.language}`);
  }
  const args = [...commandTail, harnessPath];

  const requestJson = JSON.stringify(request);

  return new Promise<InvokeResponse>((resolve) => {
    const child = spawn(cmd, args, {
      stdio: ['pipe', 'pipe', 'pipe'],
      timeout: timeoutMs,
      env: {
        ...process.env,
        GNODE_POLYGLOT: '1',
      },
    });

    let stdout = '';
    let stderr = '';
    let settled = false;

    const timer = setTimeout(() => {
      if (!settled) {
        settled = true;
        child.kill('SIGKILL');
        resolve({
          status: 'error',
          value: `subprocess timed out after ${timeoutMs}ms`,
          stdout,
          stderr,
          mappedError: mapErrorToSource(
            request.language,
            request.filePath,
            request.functionName,
            `subprocess timed out after ${timeoutMs}ms`,
            stderr,
            request.sourceRange
          ),
        });
      }
    }, timeoutMs);

    child.stdout.on('data', (chunk: Buffer) => {
      stdout += chunk.toString();
    });

    child.stderr.on('data', (chunk: Buffer) => {
      stderr += chunk.toString();
    });

    const settleWithError = (message: string): void => {
      if (!settled) {
        settled = true;
        clearTimeout(timer);
        resolve({
          status: 'error',
          value: message,
          stdout,
          stderr,
          mappedError: mapErrorToSource(
            request.language,
            request.filePath,
            request.functionName,
            message,
            stderr,
            request.sourceRange
          ),
        });
      }
    };

    child.stdin.on('error', (error: Error) => {
      settleWithError(error.message);
    });

    child.on('error', (error: Error) => {
      settleWithError(error.message);
    });

    child.on('close', (code: number | null) => {
      if (!settled) {
        settled = true;
        clearTimeout(timer);

        if (code !== 0) {
          const errorValue = `process exited with code ${code}`;
          resolve({
            status: 'error',
            value: errorValue,
            stdout,
            stderr,
            mappedError: mapErrorToSource(
              request.language,
              request.filePath,
              request.functionName,
              `${errorValue}\n${stdout}`,
              stderr,
              request.sourceRange
            ),
          });
          return;
        }

        try {
          const response = JSON.parse(stdout) as InvokeResponse;
          // Add error mapping if the harness reported an error.
          if (response.status === 'error') {
            response.mappedError = mapErrorToSource(
              request.language,
              request.filePath,
              request.functionName,
              response.value,
              response.stderr,
              request.sourceRange
            );
          }
          resolve(response);
        } catch {
          // If harness didn't produce JSON, wrap raw output.
          resolve({
            status: 'ok',
            value: stdout.trim(),
            stdout,
            stderr,
          });
        }
      }
    });

    // Send the request.
    child.stdin.write(requestJson, (error?: Error | null) => {
      if (error) {
        settleWithError(error.message);
        return;
      }
      child.stdin.end();
    });
  });
}

/**
 * Invoke with a simple function call -- convenience wrapper.
 */
export async function invokeFunction(
  language: string,
  filePath: string,
  functionName: string,
  args: unknown[] = [],
  sourceRange?: { startByte: number; endByte: number }
): Promise<InvokeResponse> {
  const request: InvokeRequest = {
    action: 'execute',
    language,
    filePath,
    functionName,
    args,
  };
  if (sourceRange !== undefined) {
    request.sourceRange = sourceRange;
  }
  return invokeSubprocess(request);
}

// ─── Streaming Invocation ────────────────────────────────────────────────────

/**
 * Callback for streaming subprocess output.
 */
export type StreamCallback = (
  chunk: string,
  stream: 'stdout' | 'stderr'
) => void;

/**
 * Invoke a subprocess with streaming output.
 * Instead of buffering all output, calls onChunk as data arrives.
 * Useful for long-running polyglot functions.
 */
export async function invokeSubprocessStreaming(
  request: InvokeRequest,
  onChunk: StreamCallback
): Promise<InvokeResponse> {
  const timeoutMs = getTimeoutMs(request);
  const { harnessPath, commandParts } = resolveInvocationHarness(request);

  const [cmd, ...commandTail] = commandParts;
  if (!cmd) {
    throw new Error(`No launch command configured for language ${request.language}`);
  }
  const args = [...commandTail, harnessPath];

  const requestJson = JSON.stringify(request);

  return new Promise<InvokeResponse>((resolve) => {
    const child = spawn(cmd, args, {
      stdio: ['pipe', 'pipe', 'pipe'],
      timeout: timeoutMs,
      env: {
        ...process.env,
        GNODE_POLYGLOT: '1',
        GNODE_STREAMING: '1',
      },
    });

    let stdout = '';
    let stderr = '';
    let settled = false;

    const timer = setTimeout(() => {
      if (!settled) {
        settled = true;
        child.kill('SIGKILL');
        resolve({
          status: 'error',
          value: `subprocess timed out after ${timeoutMs}ms`,
          stdout,
          stderr,
        });
      }
    }, timeoutMs);

    child.stdout.on('data', (chunk: Buffer) => {
      const text = chunk.toString();
      stdout += text;
      onChunk(text, 'stdout');
    });

    child.stderr.on('data', (chunk: Buffer) => {
      const text = chunk.toString();
      stderr += text;
      onChunk(text, 'stderr');
    });

    const settleWithError = (message: string): void => {
      if (!settled) {
        settled = true;
        clearTimeout(timer);
        resolve({
          status: 'error',
          value: message,
          stdout,
          stderr,
        });
      }
    };

    child.stdin.on('error', (error: Error) => {
      settleWithError(error.message);
    });

    child.on('error', (error: Error) => {
      settleWithError(error.message);
    });

    child.on('close', (code: number | null) => {
      if (!settled) {
        settled = true;
        clearTimeout(timer);

        if (code !== 0) {
          resolve({
            status: 'error',
            value: `process exited with code ${code}`,
            stdout,
            stderr,
          });
          return;
        }

        try {
          const response = JSON.parse(stdout) as InvokeResponse;
          resolve(response);
        } catch {
          resolve({
            status: 'ok',
            value: stdout.trim(),
            stdout,
            stderr,
          });
        }
      }
    });

    child.stdin.write(requestJson, (error?: Error | null) => {
      if (error) {
        settleWithError(error.message);
        return;
      }
      child.stdin.end();
    });
  });
}

// ─── Scanner Service (Persistent Mode) ──────────────────────────────────────

/**
 * Persistent polyglot scanner service.
 * Keeps the Rust gnosis-polyglot binary alive and sends requests over stdin,
 * avoiding the ~50ms startup cost per parse.
 */
export class PolyglotScannerService {
  private process: ChildProcess | null = null;
  private pending: Array<{
    resolve: (value: string) => void;
    reject: (error: Error) => void;
    buffer: string;
  }> = [];
  private buffer = '';

  private readonly binaryPath: string | undefined;

  constructor(binaryPath?: string) {
    this.binaryPath = binaryPath;
  }

  /**
   * Get or start the persistent scanner process.
   */
  private ensureProcess(): ChildProcess {
    if (this.process && !this.process.killed) {
      return this.process;
    }

    const binary = this.binaryPath ?? this.resolveBinaryPath();
    this.process = spawn(binary, ['--format', 'json', '--stdin'], {
      stdio: ['pipe', 'pipe', 'pipe'],
      env: { ...process.env },
    });

    this.process.stdout?.on('data', (chunk: Buffer) => {
      this.buffer += chunk.toString();
      this.drainBuffer();
    });

    this.process.on('error', (error: Error) => {
      for (const p of this.pending) {
        p.reject(error);
      }
      this.pending = [];
    });

    this.process.on('close', () => {
      this.process = null;
      for (const p of this.pending) {
        p.reject(new Error('scanner process exited'));
      }
      this.pending = [];
    });

    return this.process;
  }

  /**
   * Drain the buffer, resolving pending requests when complete JSON is found.
   */
  private drainBuffer(): void {
    // Look for complete JSON objects (delimited by newlines in NDJSON mode).
    const lines = this.buffer.split('\n');
    // Keep the last incomplete line in the buffer.
    this.buffer = lines.pop() ?? '';

    for (const line of lines) {
      if (line.trim().length === 0) continue;
      const pending = this.pending.shift();
      if (pending) {
        pending.resolve(line);
      }
    }
  }

  /**
   * Send a file to the scanner and get the result.
   */
  async scan(filePath: string): Promise<string> {
    const proc = this.ensureProcess();

    return new Promise<string>((resolve, reject) => {
      this.pending.push({ resolve, reject, buffer: '' });
      proc.stdin?.write(`${filePath}\n`);
    });
  }

  /**
   * Stop the scanner service.
   */
  stop(): void {
    if (this.process) {
      this.process.kill('SIGTERM');
      this.process = null;
    }
    this.pending = [];
  }

  private resolveBinaryPath(): string {
    return path.resolve(
      path.dirname(new URL(import.meta.url).pathname),
      '..',
      '..',
      'polyglot',
      'target',
      'release',
      'gnosis-polyglot'
    );
  }
}
