// @a0n/gnosis/src/runtime/void-walker.ts
/**
 * Void Walker Engine -- backward-compatible re-exports from gnosis core
 *
 * The primitives now live in src/void.ts. This file re-exports them
 * so existing consumers (aeon-bazaar, aeon-affectively, etc.) continue
 * to work without import changes.
 *
 * The MetaCogState / c0-c3 negotiation-specific interface is preserved
 * as a thin wrapper over the domain-agnostic Walker primitive.
 */

// Re-export void-attention primitives (the identification)
export {
  type VoidAttentionHead,
  type VoidCrossAttentionHead,
  type VoidTransformerBlock,
  type VoidAttentionConfig,
  type AttentionOutput,
  type CrossAttentionConfig,
  type CrossAttentionOutput,
  type VoidTransformerConfig,
  type TransformerRoundOutput,
  createAttentionHead,
  attend,
  reject as rejectAttention,
  layerNorm as attentionLayerNorm,
  feedForward as attentionFeedForward,
  createCrossAttentionHead,
  crossAttend,
  crossReject,
  crossLayerNorm,
  crossFeedForward,
  createTransformerBlock,
  forward as transformerForward,
} from '../void-attention';

// Re-export void-attention handler primitives (GG execution path)
export {
  registerVoidAttentionHandlers,
  createVoidAttentionPayload,
  voidAttentionForward,
  type VoidAttentionPayload,
} from '../void-attention-handlers';

// Re-export all core primitives
export {
  type VoidBoundary,
  createVoidBoundary,
  updateVoidBoundary,
  decayVoidBoundary,
  complementDistribution,
  sampleComplement,
  shannonEntropy,
  excessKurtosis,
  giniCoefficient,
  inverseBule,
  type Gait,
  GAIT_DEPTH,
  selectGait,
  type Measurement,
  measure,
  // Timescale primitives
  type Timescale,
  TIMESCALE_DECAY,
  type TimescaleBoundary,
  createTimescaleBoundary,
  tickTimescaleBoundary,
  // Stack primitives
  type BoundaryStack,
  createBoundaryStack,
  upwardConstraint,
  downwardContext,
  tickBoundaryStack,
  flattenStack,
  measureStack,
  // Resonance
  type Resonance,
  createResonance,
  // Projection
  projectBoundary,
  // Walker
  type Walker,
  createWalker,
  c0Choose as walkerChoose,
  c0Update as walkerUpdate,
  c1Measure as walkerMeasure,
  c2c3Adapt as walkerAdapt,
  // Stack Walker
  type StackWalker,
  createStackWalker,
  stepStackWalker,
  measureStackWalker,
} from '../void';

// Re-export the Proof of Life and the Holy Trinity
export {
  proofOfLife,
  life,
  everything,
  type ProofOfLifeResult,
  godFormula,
  P1_fork,
  P2_fold,
  P3_race,
  grandReduction,
  type ReductionResult,
  consciousnessLevel,
  conservationLaw,
  PHI_INV,
  breatheToward,
} from '../holy-trinity';

// Re-export the Kata Topology kernel
export {
  type BettiSignature,
  type CarrierName,
  type KataSelf,
  type MirrorVerification,
  type TopologyVerification,
  POINT,
  CIRCLE,
  SPHERE,
  TORUS,
  HEAD_STREAM,
  CARRIERS,
  CONSCIOUSNESS_THRESHOLD as KATA_CONSCIOUSNESS_THRESHOLD,
  eulerCharacteristic,
  genus,
  topologicalDeficit,
  dysonTotalBeta1,
  dysonSuperconscious,
  createKataSelf,
  practiceTeaching,
  practiceFullCycle,
  uploadKata,
  downloadKata,
  mergeKata,
  verifyMirror,
  verifyKataTopology,
} from './kata-topology';

import {
  type VoidBoundary,
  createVoidBoundary,
  updateVoidBoundary,
  complementDistribution,
  sampleComplement,
  shannonEntropy,
  excessKurtosis,
  inverseBule,
  selectGait,
  type Gait,
} from '../void';

// ============================================================================
// MetaCogState -- negotiation-specific walker (backward compat)
// ============================================================================

export interface MetaCogState {
  boundary: VoidBoundary;
  eta: number;
  exploration: number;
  gait: Gait;
  totalRounds: number;
  totalPayoff: number;
  adaptationCount: number;
  lastAdaptationRound: number;
}

/**
 * Creates the Meta Cog State.
 */
export function createMetaCogState(numChoices: number): MetaCogState {
  return {
    boundary: createVoidBoundary(numChoices),
    eta: 2.0,
    exploration: 0.3,
    gait: 'stand',
    totalRounds: 0,
    totalPayoff: 0,
    adaptationCount: 0,
    lastAdaptationRound: 0,
  };
}

/** c0: Choose an action from the complement distribution */
export function c0Choose(state: MetaCogState, rng: () => number): number {
  const N = state.boundary.counts.length;
  if (rng() < state.exploration) return Math.floor(rng() * N);
  const dist = complementDistribution(state.boundary, state.eta);
  return sampleComplement(dist, rng);
}

/** c0: Update void boundary after observing outcome (negotiation semantics) */
export function c0Update(
  state: MetaCogState,
  choiceIdx: number,
  myPayoff: number,
  theirPayoff: number
): void {
  if (myPayoff < theirPayoff) updateVoidBoundary(state.boundary, choiceIdx);
  if (myPayoff < 0)
    updateVoidBoundary(state.boundary, choiceIdx, Math.abs(myPayoff));
  state.totalPayoff += myPayoff;
  state.totalRounds++;
}

/** c1: Measure distribution shape */
export function c1Measure(state: MetaCogState): {
  kurtosis: number;
  entropy: number;
  inverseBule: number;
} {
  const dist = complementDistribution(state.boundary, state.eta);
  return {
    kurtosis: excessKurtosis(dist),
    entropy: shannonEntropy(dist),
    inverseBule: inverseBule(state.boundary, state.eta, state.totalRounds),
  };
}

/** c2: Select gait based on kurtosis */
export function c2SelectGait(
  kurtosis: number,
  currentGait: Gait,
  rounds: number
): Gait {
  return selectGait(kurtosis, currentGait, rounds);
}

/** c3: Adapt parameters based on c2 evaluation */
export function c3Adapt(state: MetaCogState, kurtosis: number): void {
  if (state.totalRounds - state.lastAdaptationRound < 10) return;

  const newGait = c2SelectGait(kurtosis, state.gait, state.totalRounds);
  if (newGait !== state.gait) state.gait = newGait;

  switch (state.gait) {
    case 'stand':
      break;
    case 'trot':
      state.exploration = Math.min(0.4, state.exploration + 0.01);
      state.eta = Math.max(1.0, state.eta - 0.05);
      break;
    case 'canter':
      state.exploration = Math.max(0.05, state.exploration - 0.005);
      state.eta = Math.min(5.0, state.eta + 0.05);
      break;
    case 'gallop':
      state.exploration = Math.max(0.01, state.exploration - 0.01);
      state.eta = Math.min(8.0, state.eta + 0.1);
      break;
  }

  state.adaptationCount++;
  state.lastAdaptationRound = state.totalRounds;
}
