// @a0n/.ignored_gnosis/src/runtime/agent-swarm-contract.ts
import { sha256Hex, stableStringify } from '../universe/kernel';

export type AgentSwarmDecision = 'merge-ready' | 'escalate' | 'reject';
export type AgentSwarmIssueSeverity = 'info' | 'warning' | 'error';
export type AgentSwarmRiskLevel =
  | 'standard'
  | 'critical'
  | 'security'
  | 'release'
  | 'theorem';
export type AgentSwarmExecutionStatus =
  | 'completed'
  | 'failed'
  | 'aborted'
  | 'skipped';
export type AgentSwarmVerificationStatus = 'passed' | 'failed' | 'skipped';
export type AgentSwarmReviewerType = 'human' | 'agent';

export interface AgentSwarmTaskContract {
  readonly id: string;
  readonly title: string;
  readonly owner: string;
  readonly allowedPathScopes: readonly string[];
  readonly allowedTools: readonly string[];
  readonly requiredVerifications: readonly string[];
  readonly allowPathOverlapWith: readonly string[];
  readonly requiresTrace: boolean;
  readonly requiresHumanReview: boolean;
}

export interface AgentSwarmVerificationResult {
  readonly id: string;
  readonly status: AgentSwarmVerificationStatus;
  readonly detail?: string;
}

export interface AgentSwarmTaskExecutionRecord {
  readonly taskId: string;
  readonly agentId: string;
  readonly status: AgentSwarmExecutionStatus;
  readonly touchedPaths: readonly string[];
  readonly toolsUsed: readonly string[];
  readonly traceId: string | null;
  readonly traceDigest: string | null;
  readonly verificationResults: readonly AgentSwarmVerificationResult[];
  readonly startedAt: string | null;
  readonly finishedAt: string | null;
}

export interface AgentSwarmReviewerApproval {
  readonly reviewerId: string;
  readonly reviewerType: AgentSwarmReviewerType;
  readonly approved: boolean;
  readonly role: string | null;
}

export interface AgentSwarmReviewPolicy {
  readonly minimumApprovals: number;
  readonly minimumHumanApprovals: number;
  readonly requireHumanOnProtectedBranch: boolean;
}

export interface AgentSwarmGlobalVerificationPolicy {
  readonly requiredChecks: readonly string[];
}

export interface AgentSwarmContractRequest {
  readonly changeId: string;
  readonly targetBranch: string;
  readonly protectedBranch: boolean;
  readonly riskLevel: AgentSwarmRiskLevel;
  readonly taskContracts: readonly AgentSwarmTaskContract[];
  readonly taskExecutions: readonly AgentSwarmTaskExecutionRecord[];
  readonly approvals: readonly AgentSwarmReviewerApproval[];
  readonly reviewPolicy: AgentSwarmReviewPolicy;
  readonly globalVerification: AgentSwarmGlobalVerificationPolicy;
}

export interface AgentSwarmContractIssue {
  readonly code: string;
  readonly severity: AgentSwarmIssueSeverity;
  readonly message: string;
  readonly taskId?: string;
  readonly paths?: readonly string[];
}

export interface AgentSwarmContractMetrics {
  readonly taskCount: number;
  readonly executionCount: number;
  readonly completedTaskCount: number;
  readonly completedExecutionCount: number;
  readonly uniqueTouchedPathCount: number;
  readonly overlappingPathCount: number;
  readonly unauthorizedOverlapCount: number;
  readonly missingTraceCount: number;
  readonly verificationPassCount: number;
  readonly verificationFailCount: number;
  readonly verificationMissingCount: number;
  readonly approvalCount: number;
  readonly humanApprovalCount: number;
}

export interface AgentSwarmContractWitness {
  readonly ownershipBoundariesSatisfied: boolean;
  readonly pathScopeSatisfied: boolean;
  readonly toolScopeSatisfied: boolean;
  readonly provenanceSatisfied: boolean;
  readonly verificationCustodySatisfied: boolean;
  readonly reviewerQuorumSatisfied: boolean;
  readonly overlapSatisfied: boolean;
  readonly protectedBranchSatisfied: boolean;
}

export interface AgentSwarmMergeCertificate {
  readonly version: 1;
  readonly digest: string;
  readonly generatedAt: number;
  readonly decision: AgentSwarmDecision;
  readonly changeId: string;
  readonly targetBranch: string;
  readonly protectedBranch: boolean;
  readonly riskLevel: AgentSwarmRiskLevel;
  readonly contractDigest: string;
  readonly traceDigest: string;
  readonly metrics: AgentSwarmContractMetrics;
  readonly witness: AgentSwarmContractWitness;
  readonly issues: readonly AgentSwarmContractIssue[];
  readonly snapshot: AgentSwarmContractRequest;
}

export interface AgentSwarmCertificateAudit {
  readonly selfConsistent: boolean;
  readonly issues: readonly string[];
}

export interface AgentSwarmAdjudication {
  readonly decision: AgentSwarmDecision;
  readonly issues: readonly AgentSwarmContractIssue[];
  readonly metrics: AgentSwarmContractMetrics;
  readonly witness: AgentSwarmContractWitness;
  readonly certificate: AgentSwarmMergeCertificate;
  readonly audit: AgentSwarmCertificateAudit;
}

const VALID_RISK_LEVELS: readonly AgentSwarmRiskLevel[] = [
  'standard',
  'critical',
  'security',
  'release',
  'theorem',
];
const VALID_EXECUTION_STATUSES: readonly AgentSwarmExecutionStatus[] = [
  'completed',
  'failed',
  'aborted',
  'skipped',
];
const VALID_VERIFICATION_STATUSES: readonly AgentSwarmVerificationStatus[] = [
  'passed',
  'failed',
  'skipped',
];
const VALID_REVIEWER_TYPES: readonly AgentSwarmReviewerType[] = [
  'human',
  'agent',
];

type MutableMetrics = {
  taskCount: number;
  executionCount: number;
  completedTaskCount: number;
  completedExecutionCount: number;
  uniqueTouchedPathCount: number;
  overlappingPathCount: number;
  unauthorizedOverlapCount: number;
  missingTraceCount: number;
  verificationPassCount: number;
  verificationFailCount: number;
  verificationMissingCount: number;
  approvalCount: number;
  humanApprovalCount: number;
};

type ParsedRecord = Record<string, unknown>;

function asRecord(value: unknown, context: string): ParsedRecord {
  if (value === null || typeof value !== 'object' || Array.isArray(value)) {
    throw new Error(`${context} must be an object`);
  }

  return value as ParsedRecord;
}

function asString(value: unknown, context: string): string {
  if (typeof value !== 'string' || value.trim() === '') {
    throw new Error(`${context} must be a non-empty string`);
  }

  return value;
}

function asOptionalString(value: unknown): string | null {
  if (value === undefined || value === null) {
    return null;
  }
  if (typeof value !== 'string') {
    throw new Error('optional string field must be a string when provided');
  }
  const trimmed = value.trim();
  return trimmed === '' ? null : value;
}

function asBoolean(value: unknown, context: string): boolean {
  if (typeof value !== 'boolean') {
    throw new Error(`${context} must be a boolean`);
  }

  return value;
}

function asNonNegativeInteger(value: unknown, context: string): number {
  if (typeof value !== 'number' || !Number.isInteger(value) || value < 0) {
    throw new Error(`${context} must be a non-negative integer`);
  }

  return value;
}

function asStringArray(value: unknown, context: string): string[] {
  if (!Array.isArray(value)) {
    throw new Error(`${context} must be an array`);
  }

  return value.map((entry, index) => asString(entry, `${context}[${index}]`));
}

function sortStrings(values: readonly string[]): string[] {
  return [...new Set(values)].sort((left, right) => left.localeCompare(right));
}

function normalizePath(pathValue: string): string {
  const normalized = pathValue.replaceAll('\\', '/').replace(/\/+/g, '/');
  const withoutDotPrefix = normalized.replace(/^\.\//, '');
  if (withoutDotPrefix === '/') {
    return '/';
  }
  return withoutDotPrefix.replace(/\/$/, '') || '.';
}

function splitScope(scope: string): string[] {
  const normalized = normalizePath(scope);
  if (normalized === '.' || normalized === '/') {
    return normalized === '/' ? [''] : [];
  }
  return normalized.split('/');
}

function segmentMatches(scopeSegment: string, pathSegment: string): boolean {
  if (scopeSegment === '*' || scopeSegment === pathSegment) {
    return true;
  }
  if (!scopeSegment.includes('*')) {
    return false;
  }
  // Intra-segment glob: `*` matches any run of characters within a single
  // path segment (never crosses `/`, since segments are already split).
  const escaped = scopeSegment.replaceAll(/[.+?^${}()|[\]\\]/g, '\\$&');
  const pattern = `^${escaped.replaceAll('*', '[^/]*')}$`;
  return new RegExp(pattern).test(pathSegment);
}

function matchesScopeSegments(
  pathSegments: string[],
  scopeSegments: string[]
): boolean {
  if (scopeSegments.length === 0) {
    return pathSegments.length === 0;
  }

  const [scopeHead, ...scopeTail] = scopeSegments;
  if (scopeHead === '**') {
    if (scopeTail.length === 0) {
      return true;
    }
    for (let index = 0; index <= pathSegments.length; index += 1) {
      if (matchesScopeSegments(pathSegments.slice(index), scopeTail)) {
        return true;
      }
    }
    return false;
  }
  if (scopeHead === undefined) {
    return false;
  }

  const [pathHead, ...pathTail] = pathSegments;
  if (pathHead === undefined) {
    return false;
  }
  if (!segmentMatches(scopeHead, pathHead)) {
    return false;
  }
  return matchesScopeSegments(pathTail, scopeTail);
}

/**
 * Handles the gnosis path Matches Scope workflow.
 */
export function pathMatchesScope(pathValue: string, scope: string): boolean {
  if (scope === '**' || scope === '*') {
    return true;
  }

  return matchesScopeSegments(splitScope(pathValue), splitScope(scope));
}

function isAllowedTool(
  toolName: string,
  allowedTools: readonly string[]
): boolean {
  return allowedTools.includes('*') || allowedTools.includes(toolName);
}

function classifyDefaultReviewPolicy(
  protectedBranch: boolean,
  riskLevel: AgentSwarmRiskLevel
): AgentSwarmReviewPolicy {
  let minimumApprovals = protectedBranch ? 1 : 0;
  let minimumHumanApprovals = protectedBranch ? 1 : 0;

  if (
    riskLevel === 'critical' ||
    riskLevel === 'security' ||
    riskLevel === 'release' ||
    riskLevel === 'theorem'
  ) {
    minimumApprovals = Math.max(minimumApprovals, 2);
    minimumHumanApprovals = Math.max(minimumHumanApprovals, 1);
  }

  return {
    minimumApprovals,
    minimumHumanApprovals,
    requireHumanOnProtectedBranch: protectedBranch,
  };
}

function parseRiskLevel(value: unknown): AgentSwarmRiskLevel {
  if (value === undefined) {
    return 'standard';
  }
  const riskLevel = asString(value, 'riskLevel');
  if (!VALID_RISK_LEVELS.includes(riskLevel as AgentSwarmRiskLevel)) {
    throw new Error(`riskLevel must be one of ${VALID_RISK_LEVELS.join(', ')}`);
  }
  return riskLevel as AgentSwarmRiskLevel;
}

function parseTaskContract(
  value: unknown,
  index: number
): AgentSwarmTaskContract {
  const record = asRecord(value, `taskContracts[${index}]`);

  return {
    id: asString(record.id, `taskContracts[${index}].id`),
    title: asString(record.title, `taskContracts[${index}].title`),
    owner: asString(record.owner, `taskContracts[${index}].owner`),
    allowedPathScopes: sortStrings(
      asStringArray(
        record.allowedPathScopes,
        `taskContracts[${index}].allowedPathScopes`
      ).map(normalizePath)
    ),
    allowedTools: sortStrings(
      asStringArray(record.allowedTools, `taskContracts[${index}].allowedTools`)
    ),
    requiredVerifications: sortStrings(
      record.requiredVerifications === undefined
        ? []
        : asStringArray(
            record.requiredVerifications,
            `taskContracts[${index}].requiredVerifications`
          )
    ),
    allowPathOverlapWith: sortStrings(
      record.allowPathOverlapWith === undefined
        ? []
        : asStringArray(
            record.allowPathOverlapWith,
            `taskContracts[${index}].allowPathOverlapWith`
          )
    ),
    requiresTrace:
      record.requiresTrace === undefined
        ? true
        : asBoolean(
            record.requiresTrace,
            `taskContracts[${index}].requiresTrace`
          ),
    requiresHumanReview:
      record.requiresHumanReview === undefined
        ? false
        : asBoolean(
            record.requiresHumanReview,
            `taskContracts[${index}].requiresHumanReview`
          ),
  };
}

function parseVerificationResult(
  value: unknown,
  context: string
): AgentSwarmVerificationResult {
  const record = asRecord(value, context);
  const status = asString(record.status, `${context}.status`);
  const detail = asOptionalString(record.detail);
  if (
    !VALID_VERIFICATION_STATUSES.includes(
      status as AgentSwarmVerificationStatus
    )
  ) {
    throw new Error(
      `${context}.status must be one of ${VALID_VERIFICATION_STATUSES.join(
        ', '
      )}`
    );
  }

  return {
    id: asString(record.id, `${context}.id`),
    status: status as AgentSwarmVerificationStatus,
    ...(detail !== null ? { detail } : {}),
  };
}

function parseTaskExecution(
  value: unknown,
  index: number
): AgentSwarmTaskExecutionRecord {
  const record = asRecord(value, `taskExecutions[${index}]`);
  const status = asString(record.status, `taskExecutions[${index}].status`);
  if (!VALID_EXECUTION_STATUSES.includes(status as AgentSwarmExecutionStatus)) {
    throw new Error(
      `taskExecutions[${index}].status must be one of ${VALID_EXECUTION_STATUSES.join(
        ', '
      )}`
    );
  }

  return {
    taskId: asString(record.taskId, `taskExecutions[${index}].taskId`),
    agentId: asString(record.agentId, `taskExecutions[${index}].agentId`),
    status: status as AgentSwarmExecutionStatus,
    touchedPaths: sortStrings(
      record.touchedPaths === undefined
        ? []
        : asStringArray(
            record.touchedPaths,
            `taskExecutions[${index}].touchedPaths`
          ).map(normalizePath)
    ),
    toolsUsed: sortStrings(
      record.toolsUsed === undefined
        ? []
        : asStringArray(record.toolsUsed, `taskExecutions[${index}].toolsUsed`)
    ),
    traceId: asOptionalString(record.traceId),
    traceDigest: asOptionalString(record.traceDigest),
    verificationResults: (Array.isArray(record.verificationResults)
      ? record.verificationResults
      : []
    ).map((entry, verificationIndex) =>
      parseVerificationResult(
        entry,
        `taskExecutions[${index}].verificationResults[${verificationIndex}]`
      )
    ),
    startedAt: asOptionalString(record.startedAt),
    finishedAt: asOptionalString(record.finishedAt),
  };
}

function parseApproval(
  value: unknown,
  index: number
): AgentSwarmReviewerApproval {
  const record = asRecord(value, `approvals[${index}]`);
  const reviewerType = asString(
    record.reviewerType,
    `approvals[${index}].reviewerType`
  );
  if (!VALID_REVIEWER_TYPES.includes(reviewerType as AgentSwarmReviewerType)) {
    throw new Error(
      `approvals[${index}].reviewerType must be one of ${VALID_REVIEWER_TYPES.join(
        ', '
      )}`
    );
  }

  return {
    reviewerId: asString(record.reviewerId, `approvals[${index}].reviewerId`),
    reviewerType: reviewerType as AgentSwarmReviewerType,
    approved: asBoolean(record.approved, `approvals[${index}].approved`),
    role: asOptionalString(record.role),
  };
}

/**
 * Parses the Agent Swarm Contract Request.
 */
export function parseAgentSwarmContractRequest(
  value: unknown
): AgentSwarmContractRequest {
  const record = asRecord(value, 'agent swarm contract request');
  const protectedBranch = asBoolean(record.protectedBranch, 'protectedBranch');
  const riskLevel = parseRiskLevel(record.riskLevel);
  const defaultReviewPolicy = classifyDefaultReviewPolicy(
    protectedBranch,
    riskLevel
  );

  const reviewPolicyRecord =
    record.reviewPolicy === undefined
      ? null
      : asRecord(record.reviewPolicy, 'reviewPolicy');

  const reviewPolicy: AgentSwarmReviewPolicy = {
    minimumApprovals:
      reviewPolicyRecord?.minimumApprovals === undefined
        ? defaultReviewPolicy.minimumApprovals
        : asNonNegativeInteger(
            reviewPolicyRecord.minimumApprovals,
            'reviewPolicy.minimumApprovals'
          ),
    minimumHumanApprovals:
      reviewPolicyRecord?.minimumHumanApprovals === undefined
        ? defaultReviewPolicy.minimumHumanApprovals
        : asNonNegativeInteger(
            reviewPolicyRecord.minimumHumanApprovals,
            'reviewPolicy.minimumHumanApprovals'
          ),
    requireHumanOnProtectedBranch:
      reviewPolicyRecord?.requireHumanOnProtectedBranch === undefined
        ? defaultReviewPolicy.requireHumanOnProtectedBranch
        : asBoolean(
            reviewPolicyRecord.requireHumanOnProtectedBranch,
            'reviewPolicy.requireHumanOnProtectedBranch'
          ),
  };

  const globalVerificationRecord =
    record.globalVerification === undefined
      ? null
      : asRecord(record.globalVerification, 'globalVerification');

  return normalizeAgentSwarmContractRequest({
    changeId: asString(record.changeId, 'changeId'),
    targetBranch:
      record.targetBranch === undefined
        ? 'main'
        : asString(record.targetBranch, 'targetBranch'),
    protectedBranch,
    riskLevel,
    taskContracts: normalizeTaskContracts(record.taskContracts),
    taskExecutions: normalizeTaskExecutions(record.taskExecutions),
    approvals: normalizeApprovals(record.approvals),
    reviewPolicy,
    globalVerification: {
      requiredChecks: sortStrings(
        globalVerificationRecord?.requiredChecks === undefined
          ? []
          : asStringArray(
              globalVerificationRecord.requiredChecks,
              'globalVerification.requiredChecks'
            )
      ),
    },
  });
}

function normalizeTaskContracts(
  value: unknown
): readonly AgentSwarmTaskContract[] {
  if (!Array.isArray(value)) {
    throw new Error('taskContracts must be an array');
  }
  return [...value.map((entry, index) => parseTaskContract(entry, index))].sort(
    (left, right) => left.id.localeCompare(right.id)
  );
}

function normalizeTaskExecutions(
  value: unknown
): readonly AgentSwarmTaskExecutionRecord[] {
  if (!Array.isArray(value)) {
    throw new Error('taskExecutions must be an array');
  }
  return [
    ...value.map((entry, index) => parseTaskExecution(entry, index)),
  ].sort(
    (left, right) =>
      left.taskId.localeCompare(right.taskId) ||
      left.agentId.localeCompare(right.agentId) ||
      (left.traceDigest ?? '').localeCompare(right.traceDigest ?? '')
  );
}

function normalizeApprovals(
  value: unknown
): readonly AgentSwarmReviewerApproval[] {
  if (value === undefined) {
    return [];
  }
  if (!Array.isArray(value)) {
    throw new Error('approvals must be an array');
  }
  return [...value.map((entry, index) => parseApproval(entry, index))].sort(
    (left, right) =>
      left.reviewerId.localeCompare(right.reviewerId) ||
      left.reviewerType.localeCompare(right.reviewerType) ||
      Number(left.approved) - Number(right.approved)
  );
}

/**
 * Handles the gnosis normalize Agent Swarm Contract Request workflow.
 */
export function normalizeAgentSwarmContractRequest(
  request: AgentSwarmContractRequest
): AgentSwarmContractRequest {
  const normalizedTasks = request.taskContracts.map((task) => ({
    ...task,
    allowedPathScopes: sortStrings(task.allowedPathScopes.map(normalizePath)),
    allowedTools: sortStrings(task.allowedTools),
    requiredVerifications: sortStrings(task.requiredVerifications),
    allowPathOverlapWith: sortStrings(task.allowPathOverlapWith),
  }));
  const normalizedExecutions = request.taskExecutions.map((execution) => ({
    ...execution,
    touchedPaths: sortStrings(execution.touchedPaths.map(normalizePath)),
    toolsUsed: sortStrings(execution.toolsUsed),
    verificationResults: [...execution.verificationResults].sort(
      (left, right) => left.id.localeCompare(right.id)
    ),
  }));
  const normalizedApprovals = [...request.approvals].sort(
    (left, right) =>
      left.reviewerId.localeCompare(right.reviewerId) ||
      left.reviewerType.localeCompare(right.reviewerType) ||
      Number(left.approved) - Number(right.approved)
  );

  return {
    ...request,
    taskContracts: [...normalizedTasks].sort((left, right) =>
      left.id.localeCompare(right.id)
    ),
    taskExecutions: [...normalizedExecutions].sort(
      (left, right) =>
        left.taskId.localeCompare(right.taskId) ||
        left.agentId.localeCompare(right.agentId) ||
        (left.traceDigest ?? '').localeCompare(right.traceDigest ?? '')
    ),
    approvals: normalizedApprovals,
    globalVerification: {
      requiredChecks: sortStrings(request.globalVerification.requiredChecks),
    },
  };
}

function isAgentSwarmContractRequest(
  value: unknown
): value is AgentSwarmContractRequest {
  if (value === null || typeof value !== 'object' || Array.isArray(value)) {
    return false;
  }

  const record = value as Partial<AgentSwarmContractRequest>;
  return (
    typeof record.changeId === 'string' &&
    typeof record.targetBranch === 'string' &&
    typeof record.protectedBranch === 'boolean' &&
    Array.isArray(record.taskContracts) &&
    Array.isArray(record.taskExecutions) &&
    Array.isArray(record.approvals)
  );
}

function createIssue(
  code: string,
  severity: AgentSwarmIssueSeverity,
  message: string,
  options?: {
    taskId?: string;
    paths?: readonly string[];
  }
): AgentSwarmContractIssue {
  return {
    code,
    severity,
    message,
    ...(options?.taskId !== undefined && { taskId: options.taskId }),
    ...(options?.paths !== undefined && { paths: options.paths }),
  };
}

function createEmptyMetrics(
  request: AgentSwarmContractRequest
): MutableMetrics {
  const humanApprovalCount = request.approvals.filter(
    (approval) => approval.approved && approval.reviewerType === 'human'
  ).length;
  return {
    taskCount: request.taskContracts.length,
    executionCount: request.taskExecutions.length,
    completedTaskCount: 0,
    completedExecutionCount: 0,
    uniqueTouchedPathCount: 0,
    overlappingPathCount: 0,
    unauthorizedOverlapCount: 0,
    missingTraceCount: 0,
    verificationPassCount: 0,
    verificationFailCount: 0,
    verificationMissingCount: 0,
    approvalCount: request.approvals.filter((approval) => approval.approved)
      .length,
    humanApprovalCount,
  };
}

function buildContractDigest(snapshot: AgentSwarmContractRequest): string {
  return sha256Hex(
    stableStringify({
      taskContracts: snapshot.taskContracts,
      targetBranch: snapshot.targetBranch,
      protectedBranch: snapshot.protectedBranch,
      riskLevel: snapshot.riskLevel,
      reviewPolicy: snapshot.reviewPolicy,
      globalVerification: snapshot.globalVerification,
    })
  );
}

function buildTraceDigest(snapshot: AgentSwarmContractRequest): string {
  return sha256Hex(
    stableStringify({
      taskExecutions: snapshot.taskExecutions,
      approvals: snapshot.approvals,
      changeId: snapshot.changeId,
    })
  );
}

function buildCertificateDigest(
  certificate: Omit<AgentSwarmMergeCertificate, 'digest'>
): string {
  return sha256Hex(stableStringify(certificate));
}

function completedExecutionsForTask(
  executions: readonly AgentSwarmTaskExecutionRecord[]
): AgentSwarmTaskExecutionRecord[] {
  return executions.filter((execution) => execution.status === 'completed');
}

function collectVerificationBuckets(
  executions: readonly AgentSwarmTaskExecutionRecord[]
): {
  passed: Set<string>;
  failed: Set<string>;
  skipped: Set<string>;
} {
  const passed = new Set<string>();
  const failed = new Set<string>();
  const skipped = new Set<string>();

  for (const execution of executions) {
    for (const result of execution.verificationResults) {
      if (result.status === 'passed') {
        passed.add(result.id);
      } else if (result.status === 'failed') {
        failed.add(result.id);
      } else {
        skipped.add(result.id);
      }
    }
  }

  return { passed, failed, skipped };
}

function overlapAllowed(
  left: AgentSwarmTaskContract,
  right: AgentSwarmTaskContract
): boolean {
  return (
    left.allowPathOverlapWith.includes(right.id) &&
    right.allowPathOverlapWith.includes(left.id)
  );
}

function toReadonlyMetrics(metrics: MutableMetrics): AgentSwarmContractMetrics {
  return { ...metrics };
}

function computeAgentSwarmAdjudication(
  request: AgentSwarmContractRequest
): Omit<AgentSwarmAdjudication, 'audit'> {
  const issues: AgentSwarmContractIssue[] = [];
  const metrics = createEmptyMetrics(request);
  const contractById = new Map<string, AgentSwarmTaskContract>();
  const executionsByTask = new Map<string, AgentSwarmTaskExecutionRecord[]>();
  const touchedPathOwners = new Map<string, Set<string>>();

  for (const task of request.taskContracts) {
    if (contractById.has(task.id)) {
      issues.push(
        createIssue(
          'duplicate_task_id',
          'error',
          `Task contract '${task.id}' is declared more than once.`,
          { taskId: task.id }
        )
      );
      continue;
    }
    contractById.set(task.id, task);
  }

  for (const execution of request.taskExecutions) {
    const task = contractById.get(execution.taskId);
    if (!task) {
      issues.push(
        createIssue(
          'unknown_task_execution',
          'error',
          `Execution for unknown task '${execution.taskId}' cannot be admitted.`,
          { taskId: execution.taskId, paths: execution.touchedPaths }
        )
      );
      continue;
    }

    const taskExecutions = executionsByTask.get(task.id) ?? [];
    taskExecutions.push(execution);
    executionsByTask.set(task.id, taskExecutions);

    if (execution.status !== 'completed') {
      continue;
    }

    metrics.completedExecutionCount += 1;

    if (task.requiresTrace && !execution.traceId && !execution.traceDigest) {
      metrics.missingTraceCount += 1;
      issues.push(
        createIssue(
          'missing_trace',
          'error',
          `Completed execution for task '${task.id}' is missing trace provenance.`,
          { taskId: task.id, paths: execution.touchedPaths }
        )
      );
    }

    for (const touchedPath of execution.touchedPaths) {
      if (
        !task.allowedPathScopes.some((scope) =>
          pathMatchesScope(touchedPath, scope)
        )
      ) {
        issues.push(
          createIssue(
            'path_scope_violation',
            'error',
            `Task '${task.id}' touched '${touchedPath}' outside its admitted path scope.`,
            { taskId: task.id, paths: [touchedPath] }
          )
        );
      }
      const owners = touchedPathOwners.get(touchedPath) ?? new Set<string>();
      owners.add(task.id);
      touchedPathOwners.set(touchedPath, owners);
    }

    for (const toolName of execution.toolsUsed) {
      if (!isAllowedTool(toolName, task.allowedTools)) {
        issues.push(
          createIssue(
            'tool_scope_violation',
            'error',
            `Task '${task.id}' used tool '${toolName}' outside its admitted tool scope.`,
            { taskId: task.id }
          )
        );
      }
    }

    for (const verification of execution.verificationResults) {
      if (verification.status === 'passed') {
        metrics.verificationPassCount += 1;
      } else if (verification.status === 'failed') {
        metrics.verificationFailCount += 1;
      }
    }
  }

  metrics.uniqueTouchedPathCount = touchedPathOwners.size;

  for (const task of request.taskContracts) {
    const executions = executionsByTask.get(task.id) ?? [];
    if (executions.length === 0) {
      issues.push(
        createIssue(
          'missing_execution',
          'error',
          `Task '${task.id}' has no execution records.`,
          { taskId: task.id }
        )
      );
      continue;
    }

    const completedExecutions = completedExecutionsForTask(executions);
    if (completedExecutions.length === 0) {
      issues.push(
        createIssue(
          'no_completed_execution',
          'error',
          `Task '${task.id}' did not produce a completed execution to fold.`,
          { taskId: task.id }
        )
      );
      continue;
    }

    metrics.completedTaskCount += 1;

    const verificationBuckets = collectVerificationBuckets(completedExecutions);

    for (const verificationId of task.requiredVerifications) {
      if (verificationBuckets.failed.has(verificationId)) {
        issues.push(
          createIssue(
            'failed_task_verification',
            'error',
            `Task '${task.id}' failed required verification '${verificationId}'.`,
            { taskId: task.id }
          )
        );
        continue;
      }

      if (!verificationBuckets.passed.has(verificationId)) {
        metrics.verificationMissingCount += 1;
        issues.push(
          createIssue(
            'missing_task_verification',
            'error',
            `Task '${task.id}' is missing required verification '${verificationId}'.`,
            { taskId: task.id }
          )
        );
      }
    }

    if (
      executions.some((execution) => execution.status !== 'completed') &&
      completedExecutions.length > 0
    ) {
      issues.push(
        createIssue(
          'retry_or_partial_failure',
          'info',
          `Task '${task.id}' required retries or had partial failures before a completed execution was folded.`,
          { taskId: task.id }
        )
      );
    }
  }

  for (const [touchedPath, taskIds] of touchedPathOwners.entries()) {
    if (taskIds.size <= 1) {
      continue;
    }

    metrics.overlappingPathCount += 1;
    const taskIdList = [...taskIds].sort((left, right) =>
      left.localeCompare(right)
    );
    let authorized = true;

    for (let leftIndex = 0; leftIndex < taskIdList.length; leftIndex += 1) {
      const leftTaskId = taskIdList[leftIndex];
      if (leftTaskId === undefined) {
        continue;
      }
      for (
        let rightIndex = leftIndex + 1;
        rightIndex < taskIdList.length;
        rightIndex += 1
      ) {
        const rightTaskId = taskIdList[rightIndex];
        if (rightTaskId === undefined) {
          continue;
        }
        const leftTask = contractById.get(leftTaskId);
        const rightTask = contractById.get(rightTaskId);
        if (!leftTask || !rightTask) {
          continue;
        }
        if (!overlapAllowed(leftTask, rightTask)) {
          authorized = false;
        }
      }
    }

    if (!authorized) {
      metrics.unauthorizedOverlapCount += 1;
      const ownerSet = new Set(
        taskIdList
          .map((taskId) => contractById.get(taskId)?.owner)
          .filter((owner): owner is string => owner !== undefined)
      );
      issues.push(
        createIssue(
          'unauthorized_overlap',
          ownerSet.size <= 1 || !request.protectedBranch ? 'warning' : 'error',
          `Path '${touchedPath}' was touched by multiple tasks without mutual overlap permission: ${taskIdList.join(
            ', '
          )}.`,
          { paths: [touchedPath] }
        )
      );
    }
  }

  const globalVerificationBuckets = collectVerificationBuckets(
    request.taskExecutions.filter(
      (execution) => execution.status === 'completed'
    )
  );
  for (const verificationId of request.globalVerification.requiredChecks) {
    if (globalVerificationBuckets.failed.has(verificationId)) {
      issues.push(
        createIssue(
          'failed_global_verification',
          'error',
          `Global verification '${verificationId}' failed.`
        )
      );
      continue;
    }

    if (!globalVerificationBuckets.passed.has(verificationId)) {
      metrics.verificationMissingCount += 1;
      issues.push(
        createIssue(
          'missing_global_verification',
          'error',
          `Global verification '${verificationId}' is missing from the folded execution set.`
        )
      );
    }
  }

  if (request.reviewPolicy.minimumApprovals > metrics.approvalCount) {
    issues.push(
      createIssue(
        'insufficient_review_quorum',
        'warning',
        `Approved reviewer count ${metrics.approvalCount} is below the required quorum ${request.reviewPolicy.minimumApprovals}.`
      )
    );
  }

  if (
    request.reviewPolicy.minimumHumanApprovals > metrics.humanApprovalCount ||
    (request.protectedBranch &&
      request.reviewPolicy.requireHumanOnProtectedBranch &&
      metrics.humanApprovalCount === 0)
  ) {
    issues.push(
      createIssue(
        'missing_human_approval',
        'warning',
        `Human approvals ${metrics.humanApprovalCount} do not satisfy the required human quorum ${request.reviewPolicy.minimumHumanApprovals}.`
      )
    );
  }

  if (
    request.taskContracts.some((task) => task.requiresHumanReview) &&
    metrics.humanApprovalCount === 0
  ) {
    issues.push(
      createIssue(
        'human_review_required',
        'warning',
        'At least one task requires human review but no human approval is present.'
      )
    );
  }

  const issueCounts = issues.reduce(
    (counts, issue) => {
      counts[issue.severity] += 1;
      return counts;
    },
    { info: 0, warning: 0, error: 0 }
  );

  const witness: AgentSwarmContractWitness = {
    ownershipBoundariesSatisfied:
      !issues.some((issue) => issue.code === 'missing_execution') &&
      metrics.unauthorizedOverlapCount === 0,
    pathScopeSatisfied: !issues.some(
      (issue) => issue.code === 'path_scope_violation'
    ),
    toolScopeSatisfied: !issues.some(
      (issue) => issue.code === 'tool_scope_violation'
    ),
    provenanceSatisfied:
      metrics.missingTraceCount === 0 &&
      !issues.some((issue) => issue.code === 'unknown_task_execution'),
    verificationCustodySatisfied:
      metrics.verificationFailCount === 0 &&
      metrics.verificationMissingCount === 0 &&
      !issues.some(
        (issue) =>
          issue.code === 'failed_task_verification' ||
          issue.code === 'missing_task_verification' ||
          issue.code === 'failed_global_verification' ||
          issue.code === 'missing_global_verification'
      ),
    reviewerQuorumSatisfied:
      metrics.approvalCount >= request.reviewPolicy.minimumApprovals &&
      metrics.humanApprovalCount >= request.reviewPolicy.minimumHumanApprovals,
    overlapSatisfied: metrics.unauthorizedOverlapCount === 0,
    protectedBranchSatisfied:
      !request.protectedBranch ||
      metrics.humanApprovalCount > 0 ||
      !request.reviewPolicy.requireHumanOnProtectedBranch,
  };

  const decision: AgentSwarmDecision =
    issueCounts.error > 0
      ? 'reject'
      : issueCounts.warning > 0
      ? 'escalate'
      : 'merge-ready';

  const snapshot = normalizeAgentSwarmContractRequest(request);
  const metricsView = toReadonlyMetrics(metrics);
  const certificateWithoutDigest: Omit<AgentSwarmMergeCertificate, 'digest'> = {
    version: 1,
    generatedAt: Date.now(),
    decision,
    changeId: snapshot.changeId,
    targetBranch: snapshot.targetBranch,
    protectedBranch: snapshot.protectedBranch,
    riskLevel: snapshot.riskLevel,
    contractDigest: buildContractDigest(snapshot),
    traceDigest: buildTraceDigest(snapshot),
    metrics: metricsView,
    witness,
    issues,
    snapshot,
  };
  const certificate: AgentSwarmMergeCertificate = {
    ...certificateWithoutDigest,
    digest: buildCertificateDigest(certificateWithoutDigest),
  };

  return {
    decision,
    issues,
    metrics: metricsView,
    witness,
    certificate,
  };
}

/**
 * Handles the gnosis evaluate Agent Swarm Contract workflow.
 */
export function evaluateAgentSwarmContract(
  input: AgentSwarmContractRequest | unknown
): AgentSwarmAdjudication {
  const request = isAgentSwarmContractRequest(input)
    ? normalizeAgentSwarmContractRequest(input)
    : parseAgentSwarmContractRequest(input);
  const adjudication = computeAgentSwarmAdjudication(request);
  const audit = auditAgentSwarmMergeCertificate(adjudication.certificate);

  return {
    ...adjudication,
    audit,
  };
}

/**
 * Handles the gnosis audit Agent Swarm Merge Certificate workflow.
 */
export function auditAgentSwarmMergeCertificate(
  certificate: AgentSwarmMergeCertificate
): AgentSwarmCertificateAudit {
  const issues: string[] = [];
  if (certificate.version !== 1) {
    issues.push(`Unknown certificate version ${certificate.version}`);
  }

  const expectedContractDigest = buildContractDigest(certificate.snapshot);
  if (certificate.contractDigest !== expectedContractDigest) {
    issues.push('Contract digest does not match certificate snapshot');
  }

  const expectedTraceDigest = buildTraceDigest(certificate.snapshot);
  if (certificate.traceDigest !== expectedTraceDigest) {
    issues.push('Trace digest does not match certificate snapshot');
  }

  const { digest: _ignoredDigest, ...withoutDigest } = certificate;
  const expectedDigest = buildCertificateDigest(withoutDigest);
  if (certificate.digest !== expectedDigest) {
    issues.push('Certificate digest does not match certificate payload');
  }

  const expectedAdjudication = computeAgentSwarmAdjudication(
    normalizeAgentSwarmContractRequest(certificate.snapshot)
  );
  const expectedMetrics = expectedAdjudication.metrics;
  if (
    stableStringify(certificate.metrics) !== stableStringify(expectedMetrics)
  ) {
    issues.push('Certificate metrics do not match the recomputed adjudication');
  }

  const expectedWitness = expectedAdjudication.witness;
  if (
    stableStringify(certificate.witness) !== stableStringify(expectedWitness)
  ) {
    issues.push(
      'Certificate witness does not match the recomputed adjudication'
    );
  }

  if (
    stableStringify(certificate.issues) !==
    stableStringify(expectedAdjudication.issues)
  ) {
    issues.push('Certificate issues do not match the recomputed adjudication');
  }

  const expectedDecision = expectedAdjudication.decision;
  if (certificate.decision !== expectedDecision) {
    issues.push(
      'Certificate decision does not match the recomputed adjudication'
    );
  }

  if (
    certificate.decision === 'merge-ready' &&
    certificate.issues.some((issue) => issue.severity !== 'info')
  ) {
    issues.push(
      'Merge-ready certificates cannot carry warning or error issues'
    );
  }

  if (
    certificate.decision === 'escalate' &&
    certificate.issues.some((issue) => issue.severity === 'error')
  ) {
    issues.push('Escalate certificates cannot carry error issues');
  }

  return {
    selfConsistent: issues.length === 0,
    issues,
  };
}

export const auditAgentSwarmCertificate = auditAgentSwarmMergeCertificate;
