// @a0n/.ignored_gnosis/src/runtime/ambient-scheduler.ts
export interface GnosisAmbientRuntimeEntropyInput {
  readonly foldedBytes?: Uint8Array | readonly number[];
  readonly cfRay?: string | null;
  readonly colo?: string | null;
  readonly edgeColoCode?: string | null;
  readonly asNumber?: number | null;
  readonly clientTcpRtt?: number | null;
  readonly runtimeId?: string | null;
  readonly timestampMs?: number;
  readonly extraSources?: readonly string[];
}

export interface GnosisAmbientRuntimeScheduleInput
  extends GnosisAmbientRuntimeEntropyInput {
  readonly workItems: number;
  readonly memoryBudgetBytes: number;
  readonly bytesPerItem: number;
  readonly maxParallelism?: number;
  readonly edgeworkBoostUnits?: number;
  readonly includeExecutionPermutation?: boolean;
}

export interface GnosisAmbientRuntimeSchedule {
  readonly kind: 'gnosis-ambient-runtime-schedule';
  readonly version: 1;
  readonly entropyWord: number;
  readonly startOffset: number;
  readonly stride: number;
  readonly parallelism: number;
  readonly edgeworkBoostUnits: number;
  readonly edgeworkBoostLanes: number;
  readonly maxItemsPerChunk: number;
  readonly memoryBudgetBytes: number;
  readonly bytesPerItem: number;
  readonly sourceFamilies: readonly string[];
  readonly executionOrder: Uint32Array;
  readonly inverseOrder: Uint32Array;
  readonly outputIndices: Uint32Array;
}

export interface GnosisStableExecutionPermutation {
  readonly order: Uint32Array;
  readonly inverse: Uint32Array;
  readonly outputIndices: Uint32Array;
}

export const GNOSIS_AMBIENT_RUNTIME_SOURCE_FAMILIES = Object.freeze([
  'folded-bytes',
  'cf-ray',
  'network-weather',
  'edge-colo',
  'runtime-id',
  'request-time',
  'edgework-boost',
] as const);

/**
 * Build a memory-first runtime schedule from ambient entropy.
 *
 * The entropy does not duplicate branches or allocate alternate tensors. It
 * only chooses a permutation and bounded parallelism under the supplied byte
 * budget, so WASM, gnode, polyglot, and worker runtimes can vary work order
 * while keeping peak live memory controlled.
 */
export function planGnosisAmbientRuntimeSchedule(
  input: GnosisAmbientRuntimeScheduleInput
): GnosisAmbientRuntimeSchedule {
  const workItems = positiveInteger(input.workItems, 'workItems');
  const bytesPerItem = positiveInteger(input.bytesPerItem, 'bytesPerItem');
  const memoryBudgetBytes = positiveInteger(
    input.memoryBudgetBytes,
    'memoryBudgetBytes'
  );
  const maxItemsByMemory = Math.max(1, Math.floor(memoryBudgetBytes / bytesPerItem));
  const maxParallelism = Math.max(
    1,
    Math.floor(input.maxParallelism ?? maxItemsByMemory)
  );
  const entropyWord = foldGnosisAmbientRuntimeEntropy(input);
  const memoryParallelismCap = Math.max(
    1,
    Math.min(maxParallelism, maxItemsByMemory)
  );
  const baseParallelism = 1 + (entropyWord % memoryParallelismCap);
  const edgeworkBoostUnits = normalizeEdgeworkBoostUnits(
    input.edgeworkBoostUnits
  );
  const edgeworkBoostLanes = Math.min(
    memoryParallelismCap - baseParallelism,
    edgeworkBoostUnitsToLanes(edgeworkBoostUnits)
  );
  const parallelism = baseParallelism + edgeworkBoostLanes;
  const maxItemsPerChunk = Math.max(
    1,
    Math.min(workItems, Math.floor(maxItemsByMemory / parallelism))
  );
  const startOffset = entropyWord % workItems;
  const stride = coprimeStride(workItems, entropyWord >>> 1);
  const permutation = input.includeExecutionPermutation
    ? planGnosisStableExecutionPermutation({
        workItems,
        startOffset,
        stride,
      })
    : EMPTY_EXECUTION_PERMUTATION;

  return {
    kind: 'gnosis-ambient-runtime-schedule',
    version: 1,
    entropyWord,
    startOffset,
    stride,
    parallelism,
    edgeworkBoostUnits,
    edgeworkBoostLanes,
    maxItemsPerChunk,
    memoryBudgetBytes,
    bytesPerItem,
    sourceFamilies: GNOSIS_AMBIENT_RUNTIME_SOURCE_FAMILIES,
    executionOrder: permutation.order,
    inverseOrder: permutation.inverse,
    outputIndices: permutation.outputIndices,
  };
}

/**
 * Handles the gnosis plan Gnosis Stable Execution Permutation workflow.
 */
export function planGnosisStableExecutionPermutation(input: {
  readonly workItems: number;
  readonly startOffset?: number;
  readonly stride?: number;
  readonly costs?: Uint32Array | readonly number[];
  readonly corridors?: Uint32Array | readonly number[];
}): GnosisStableExecutionPermutation {
  const workItems = positiveInteger(input.workItems, 'workItems');
  const startOffset = input.startOffset ?? 0;
  const stride = input.stride ?? 1;
  const orderValues: number[] = [];
  const ordinalByIndex = new Map<number, number>();

  for (let ordinal = 0; ordinal < workItems; ordinal += 1) {
    const index = (startOffset + ordinal * stride) % workItems;
    orderValues.push(index);
    ordinalByIndex.set(index, ordinal);
  }

  orderValues.sort((left, right) => {
    const costDelta = valueAt(input.costs, left) - valueAt(input.costs, right);
    if (costDelta !== 0) return costDelta;
    const corridorDelta = valueAt(input.corridors, left) - valueAt(input.corridors, right);
    if (corridorDelta !== 0) return corridorDelta;
    return (ordinalByIndex.get(left) ?? left) - (ordinalByIndex.get(right) ?? right);
  });

  const order = Uint32Array.from(orderValues);
  const inverse = new Uint32Array(workItems);
  for (let slot = 0; slot < order.length; slot += 1) {
    const itemIndex = order[slot];
    if (itemIndex === undefined) {
      throw new Error('Ambient scheduler order is missing an item index');
    }
    inverse[itemIndex] = slot;
  }

  return {
    order,
    inverse,
    outputIndices: Uint32Array.from(order),
  };
}

/**
 * Handles the gnosis fold Gnosis Ambient Runtime Entropy workflow.
 */
export function foldGnosisAmbientRuntimeEntropy(
  input: GnosisAmbientRuntimeEntropyInput
): number {
  let state = 0x811c_9dc5;
  if (input.foldedBytes) {
    state = foldBytesIntoWord(state, input.foldedBytes);
  }
  const source = [
    input.cfRay,
    input.colo,
    input.edgeColoCode,
    input.asNumber,
    input.clientTcpRtt,
    input.runtimeId,
    input.timestampMs ?? Date.now(),
    ...(input.extraSources ?? []),
  ].join('\x1f');
  return foldStringIntoWord(state, source);
}

/**
 * Handles the gnosis gnosis Ambient Runtime Work Index workflow.
 */
export function gnosisAmbientRuntimeWorkIndex(
  schedule: Pick<GnosisAmbientRuntimeSchedule, 'startOffset' | 'stride'>,
  ordinal: number,
  workItems: number
): number {
  const count = positiveInteger(workItems, 'workItems');
  const index = positiveInteger(ordinal + 1, 'ordinal') - 1;
  return (schedule.startOffset + index * schedule.stride) % count;
}

/**
 * Handles the gnosis gnosis Ambient Runtime Work Order workflow.
 */
export function* gnosisAmbientRuntimeWorkOrder(
  schedule: Pick<GnosisAmbientRuntimeSchedule, 'startOffset' | 'stride'>,
  workItems: number
): Generator<number> {
  const executionOrder = 'executionOrder' in schedule
    ? (schedule as Pick<GnosisAmbientRuntimeSchedule, 'executionOrder'>).executionOrder
    : undefined;
  if (executionOrder && executionOrder.length === workItems) {
    for (const index of executionOrder) {
      yield index;
    }
    return;
  }
  for (let ordinal = 0; ordinal < workItems; ordinal += 1) {
    yield gnosisAmbientRuntimeWorkIndex(schedule, ordinal, workItems);
  }
}

function foldBytesIntoWord(
  initial: number,
  bytes: Uint8Array | readonly number[]
): number {
  let state = initial >>> 0;
  for (let i = 0; i < bytes.length; i += 1) {
    const byte = bytes[i];
    if (byte === undefined) {
      throw new Error('Ambient scheduler encountered an undefined byte');
    }
    state ^= byte & 0xff;
    state = Math.imul(state, 0x0100_0193) >>> 0;
    state ^= state >>> 13;
  }
  return state >>> 0;
}

function foldStringIntoWord(initial: number, value: string): number {
  let state = initial >>> 0;
  for (let i = 0; i < value.length; i += 1) {
    state ^= value.charCodeAt(i) & 0xff;
    state = Math.imul(state, 0x0100_0193) >>> 0;
    state ^= state >>> 13;
  }
  return state >>> 0;
}

function coprimeStride(workItems: number, seed: number): number {
  if (workItems === 1) return 1;
  let stride = (seed % (workItems - 1)) + 1;
  while (gcd(stride, workItems) !== 1) {
    stride = (stride % (workItems - 1)) + 1;
  }
  return stride;
}

function gcd(a: number, b: number): number {
  let x = Math.abs(a);
  let y = Math.abs(b);
  while (y !== 0) {
    const next = x % y;
    x = y;
    y = next;
  }
  return x;
}

function normalizeEdgeworkBoostUnits(value: number | undefined): number {
  if (typeof value !== 'number' || !Number.isFinite(value) || value <= 0) {
    return 0;
  }
  return Math.min(1_000_000, Math.floor(value * 1_000) / 1_000);
}

function edgeworkBoostUnitsToLanes(edgeworkBoostUnits: number): number {
  if (edgeworkBoostUnits <= 0) {
    return 0;
  }
  return Math.max(1, Math.floor(Math.log2(1 + edgeworkBoostUnits)));
}

function positiveInteger(value: number, name: string): number {
  if (!Number.isInteger(value) || value < 1) {
    throw new RangeError(`${name} must be a positive integer`);
  }
  return value;
}

function valueAt(values: Uint32Array | readonly number[] | undefined, index: number): number {
  return values?.[index] ?? 0;
}

const EMPTY_EXECUTION_PERMUTATION: GnosisStableExecutionPermutation = Object.freeze({
  order: new Uint32Array(0),
  inverse: new Uint32Array(0),
  outputIndices: new Uint32Array(0),
});
