// @a0n/.ignored_gnosis/src/runtime/ambient-substrate.test.ts
import { describe, expect, it } from '@a0n/gnosis/test';
import {
  GNOSIS_AMBIENT_SUBSTRATES,
  gnosisAmbientSubstrateEnv,
  gnosisAmbientSubstrateWorkOrder,
  planGnosisAmbientSubstrateSchedule,
} from './ambient-substrate';

describe('planGnosisAmbientSubstrateSchedule', () => {
  it('covers all current Gnosis substrate entrypoints', () => {
    expect(GNOSIS_AMBIENT_SUBSTRATES).toEqual([
      'polyglot',
      'gnode',
      'gnexec',
      'monster',
      'moonshine',
    ]);
  });

  it('creates a memory-first schedule for every substrate', () => {
    for (const substrate of GNOSIS_AMBIENT_SUBSTRATES) {
      const schedule = planGnosisAmbientSubstrateSchedule({
        substrate,
        foldedBytes: Uint8Array.from([7, 10, 13, 21, 34, 55]),
        cfRay: '868875887cf72f3a-SJC',
        colo: 'SJC',
        clientTcpRtt: 19,
        timestampMs: 1_765_000_000_000,
        workItems: 128,
        memoryBudgetBytes: 8192,
        bytesPerItem: 256,
        maxParallelism: 64,
      });

      expect(schedule.kind).toBe('gnosis-ambient-substrate-schedule');
      expect(schedule.substrate).toBe(substrate);
      expect(schedule.substrateRuntimeId).toBe(`gnosis://${substrate}/ambient`);
      expect(schedule.admission).toEqual({
        duplicatePayloads: false,
        memoryFirst: true,
        cryptographic: false,
      });
      expect(schedule.parallelism).toBeLessThanOrEqual(32);
      expect(schedule.maxItemsPerChunk * schedule.parallelism * 256).toBeLessThanOrEqual(
        8192
      );
    }
  });

  it('exports a CLI-friendly environment handoff', () => {
    const schedule = planGnosisAmbientSubstrateSchedule({
      substrate: 'gnexec',
      runtimeId: 'gnexec://hot-loop',
      foldedBytes: new Uint8Array(32).fill(3),
      workItems: 16,
      memoryBudgetBytes: 2048,
      bytesPerItem: 128,
      maxParallelism: 8,
      edgeworkBoostUnits: 4.25,
    });
    const env = gnosisAmbientSubstrateEnv(schedule);

    expect(env.GNOSIS_AMBIENT_SUBSTRATE).toBe('gnexec');
    expect(env.GNOSIS_AMBIENT_RUNTIME_ID).toBe('gnexec://hot-loop');
    expect(env.GNOSIS_AMBIENT_PARALLELISM).toBe(String(schedule.parallelism));
    expect(env.GNOSIS_EDGEWORK_BOOST_UNITS).toBe('4.25');
    expect(env.GNOSIS_EDGEWORK_BOOST_LANES).toBe(
      String(schedule.edgeworkBoostLanes)
    );
    expect(env.GNOSIS_AMBIENT_MAX_ITEMS_PER_CHUNK).toBe(
      String(schedule.maxItemsPerChunk)
    );
    expect(env.GNOSIS_AMBIENT_EXECUTION_PERMUTATION).toBe('0');
    expect(Number(env.GNOSIS_MONSTER_RESIDENT_WINDOW)).toBeGreaterThanOrEqual(1);
  });

  it('keeps substrate work order as a full permutation', () => {
    const schedule = planGnosisAmbientSubstrateSchedule({
      substrate: 'moonshine',
      foldedBytes: Uint8Array.from([1, 1, 2, 3, 5, 8, 13]),
      workItems: 37,
      memoryBudgetBytes: 4096,
      bytesPerItem: 128,
      maxParallelism: 12,
    });
    const order = gnosisAmbientSubstrateWorkOrder(schedule, 37);

    expect(new Set(order).size).toBe(37);
    expect(order.every((index) => index >= 0 && index < 37)).toBe(true);
  });
});
