// @a0n/.ignored_gnosis/src/runtime/compiled-topology.ts
/**
 * Compiled Topology — AOT-compiled gnosis execution
 *
 * Eliminates the engine interpretation loop entirely. At build time,
 * the gnosis AST is compiled into a flat async function that calls
 * handlers directly in topological order. No AST traversal, no
 * handler lookup, no edge filtering, no visited set, no log construction.
 *
 * The compilation pipeline:
 *   1. TS source → compileTypeScriptToGnosis() → GnosisTypeScriptBridgeResult
 *   2. GnosisTypeScriptBridgeResult → compileTopologyToFunction() → CompiledTopology
 *   3. CompiledTopology.execute(input) → result (no engine)
 *
 * This is the "cannon rubber band" pattern:
 *   - Build time: compile everything to flat functions (potential energy)
 *   - Deploy time: store in D1/KV (armed)
 *   - Cold start: preload all functions into memory (launch velocity)
 *   - Runtime: cannon fires preloaded functions through lanes (kinetic)
 *
 * The hella-whipped Wallington rotation applies at the lane level:
 * multiple CompiledTopology instances rotate through lanes with the
 * cannon cursor, each pre-armed and ready to fire.
 *
 * Performance: eliminates ~230μs engine overhead per execution,
 * leaving only handler execution time + function call overhead (~1-5μs).
 *
 * @module compiled-topology
 */

import type { GraphAST, ASTEdge, ASTNode } from '../betty/compiler';
import type { GnosisRegistry, GnosisHandler } from './registry';
import {
  buleyeanWeight,
  type VoidBoundary,
  createVoidBoundary,
  updateVoidBoundary,
  complementDistribution,
  //   sampleComplement,
} from '../void';

// ============================================================================
// Types
// ============================================================================

/**
 * A compiled execution step — one handler call with static args.
 */
export interface CompiledStep {
  /** Node ID in the original topology */
  readonly nodeId: string;
  /** The handler function to call */
  readonly handler: GnosisHandler;
  /** Static properties from the AST node */
  readonly properties: Record<string, string>;
  /** Node labels for context */
  readonly labels: readonly string[];
}

/**
 * A compiled fork — parallel branches that race or fold.
 */
export interface CompiledFork {
  readonly kind: 'fork';
  /** Branches to execute in parallel */
  readonly branches: CompiledStep[][];
  /** How to combine: 'race' (first wins) or 'fold' (merge all) */
  readonly strategy: 'race' | 'fold';
  /** Fold properties if strategy is 'fold' */
  readonly foldProps?: Record<string, string>;
}

/**
 * A compiled topology — ready to execute without the engine.
 */
export interface CompiledTopology {
  /** Sequential steps in topological order */
  readonly steps: ReadonlyArray<CompiledStep | CompiledFork>;
  /** Total node count */
  readonly nodeCount: number;
  /** Whether this topology has any forks */
  readonly hasForks: boolean;
  /** Source topology fingerprint for cache keying */
  readonly fingerprint: string;
}

/**
 * Result of compiled topology execution.
 */
export interface CompiledExecutionResult {
  /** Final payload */
  readonly payload: unknown;
  /** Execution time in microseconds */
  readonly executionUs: number;
  /** Number of handler calls made */
  readonly handlerCalls: number;
}

export interface CompiledSyncForkTopology {
  readonly steps: ReadonlyArray<CompiledStep | CompiledFork>;
  readonly nodeCount: number;
  readonly hasForks: boolean;
  readonly fingerprint: string;
}

// ============================================================================
// Compilation
// ============================================================================

/**
 * Compile a gnosis AST + registry into a flat execution plan.
 *
 * Resolves handlers at compile time so the hot path is just
 * sequential function calls — no Map lookups, no string matching.
 */
export function compileTopology(
  ast: GraphAST,
  registry: GnosisRegistry,
  fingerprint: string
): CompiledTopology {
  // Find root nodes (not targeted by any edge)
  const allTargetIds = new Set<string>();
  for (const edge of ast.edges) {
    for (const id of edge.targetIds) {
      allTargetIds.add(id.trim());
    }
  }
  const roots = [...ast.nodes.keys()].filter(
    (id) => !allTargetIds.has(id.trim())
  );
  if (roots.length === 0 && ast.nodes.size > 0) {
    const firstEdge = ast.edges[0];
    if (firstEdge) {
      const firstSourceId = firstEdge.sourceIds[0];
      if (firstSourceId) roots.push(firstSourceId.trim());
      else {
        const firstNodeId = [...ast.nodes.keys()][0];
        if (firstNodeId) roots.push(firstNodeId);
      }
    } else {
      const firstNodeId = [...ast.nodes.keys()][0];
      if (firstNodeId) roots.push(firstNodeId);
    }
  }

  // Build adjacency list
  const outEdges = new Map<string, ASTEdge[]>();
  for (const edge of ast.edges) {
    for (const sourceId of edge.sourceIds) {
      const trimmed = sourceId.trim();
      const existing = outEdges.get(trimmed) ?? [];
      existing.push(edge);
      outEdges.set(trimmed, existing);
    }
  }

  // Walk the topology in execution order, compiling steps
  const steps: Array<CompiledStep | CompiledFork> = [];
  const visited = new Set<string>();
  let hasForks = false;

  function resolveHandler(node: ASTNode): GnosisHandler | null {
    // Try each label as a handler name
    for (const label of node.labels) {
      const handler = registry.getHandler(label);
      if (handler) return handler;
    }
    // Try the node ID
    return registry.getHandler(node.id) ?? null;
  }

  function compileNode(nodeId: string): CompiledStep | null {
    const trimmed = nodeId.trim();
    if (visited.has(trimmed)) return null;
    visited.add(trimmed);

    const node = ast.nodes.get(trimmed);
    if (!node) return null;

    const handler = resolveHandler(node);
    if (!handler) return null;

    return {
      nodeId: trimmed,
      handler,
      properties: { ...node.properties },
      labels: [...node.labels],
    };
  }

  // Walk from root, following edges
  let currentNodeId: string | undefined = roots[0];

  while (currentNodeId) {
    const step = compileNode(currentNodeId);
    if (step) steps.push(step);

    const edges = outEdges.get(currentNodeId.trim()) ?? [];
    if (edges.length === 0) break;

    // Check for FORK edges
    const forkEdge = edges.find(
      (e) =>
        e.type === 'FORK' ||
        (e.properties as Record<string, unknown>)?.type === 'FORK'
    );

    if (forkEdge && forkEdge.targetIds.length > 1) {
      hasForks = true;

      // Buleyean treeshake: filter fork branches by weight.
      // Nodes with weight=1 (floor, maximally rejected) are skipped
      // if at least one non-floor branch exists. This preserves Law 1
      // (impossibility of zero) while eliminating dead branches.
      let targetIds = forkEdge.targetIds;
      const liveTargetIds = targetIds.filter((tid) => {
        const node = ast.nodes.get(tid.trim());
        if (!node) return true; // unknown = keep
        const v = parseInt(
          node.properties.rejections ?? node.properties.void_count ?? '',
          10
        );
        const r = parseInt(
          node.properties.rounds ?? node.properties.total_rounds ?? '',
          10
        );
        if (isNaN(v) || isNaN(r) || r <= 0) return true; // no data = keep
        return buleyeanWeight(r, v) > 1; // keep if above floor
      });
      // Only treeshake if we keep at least one branch (Law 1)
      if (liveTargetIds.length > 0 && liveTargetIds.length < targetIds.length) {
        targetIds = liveTargetIds;
      }

      // Compile each branch
      const branches: CompiledStep[][] = [];
      for (const targetId of targetIds) {
        const branchStep = compileNode(targetId.trim());
        if (branchStep) branches.push([branchStep]);
      }

      // Find the fold/race edge from the fork targets
      const foldTargets = new Set<string>();
      for (const targetId of forkEdge.targetIds) {
        const foldEdges = outEdges.get(targetId.trim()) ?? [];
        for (const fe of foldEdges) {
          for (const foldTarget of fe.targetIds) {
            foldTargets.add(foldTarget.trim());
          }
        }
      }

      const raceEdge = edges.find(
        (e) =>
          e.type === 'RACE' ||
          (e.properties as Record<string, unknown>)?.type === 'RACE'
      );

      steps.push({
        kind: 'fork',
        branches,
        strategy: raceEdge ? 'race' : 'fold',
        foldProps: forkEdge.properties as Record<string, string>,
      });

      // Continue from fold target
      currentNodeId = foldTargets.size > 0 ? [...foldTargets][0] : undefined;
      continue;
    }

    // Sequential: follow first edge to first target
    const nextEdge = edges[0];
    currentNodeId = nextEdge?.targetIds[0]?.trim();
  }

  return {
    steps,
    nodeCount: ast.nodes.size,
    hasForks,
    fingerprint,
  };
}

// ============================================================================
// Execution — the hot path (no engine, no AST, no lookups)
// ============================================================================

/**
 * Execute a compiled topology. This is the hot path.
 *
 * No AST traversal. No handler lookup. No edge filtering.
 * No visited set. No log construction. Just function calls.
 */
export async function executeCompiled(
  topology: CompiledTopology,
  input: unknown
): Promise<CompiledExecutionResult> {
  const start = performance.now();
  let payload = input;
  let handlerCalls = 0;

  for (const step of topology.steps) {
    if ('kind' in step && step.kind === 'fork') {
      const branchPayloads =
        Array.isArray(payload) && payload.length === step.branches.length
          ? payload
          : step.branches.map(() => payload);
      // Parallel execution
      if (step.strategy === 'race') {
        // Race: first to complete wins
        const promises = step.branches.map(async (branch, branchIndex) => {
          let branchPayload = branchPayloads[branchIndex] ?? null;
          for (const branchStep of branch) {
            branchPayload = await branchStep.handler(
              branchPayload,
              branchStep.properties,
              { nodeId: branchStep.nodeId }
            );
            handlerCalls++;
          }
          return branchPayload;
        });
        payload = await Promise.race(promises);
      } else {
        // Fold: run all, merge results
        const promises = step.branches.map(async (branch, branchIndex) => {
          let branchPayload = branchPayloads[branchIndex] ?? null;
          for (const branchStep of branch) {
            branchPayload = await branchStep.handler(
              branchPayload,
              branchStep.properties,
              { nodeId: branchStep.nodeId }
            );
            handlerCalls++;
          }
          return branchPayload;
        });
        const results = await Promise.all(promises);
        payload = mergeCompiledFoldResults(results, step.branches);
      }
    } else {
      // Sequential step — direct handler call
      const seqStep = step as CompiledStep;
      payload = await seqStep.handler(payload, seqStep.properties, {
        nodeId: seqStep.nodeId,
      });
      handlerCalls++;
    }
  }

  const executionUs = (performance.now() - start) * 1000;

  return {
    payload,
    executionUs,
    handlerCalls,
  };
}

function mergeCompiledFoldResults(
  values: unknown[],
  branches: readonly (readonly CompiledStep[])[]
): unknown {
  if (
    values.length > 0 &&
    values.every(
      (value) =>
        Array.isArray(value) &&
        value.every((entry) => typeof entry === 'number')
    )
  ) {
    const numericVectors = values as number[][];
    return numericVectors[0]!.map((_, index) =>
      numericVectors.reduce((sum, vector) => sum + (vector[index] ?? 0), 0)
    );
  }

  const merged: Record<string, unknown> = {};
  for (let i = 0; i < values.length; i++) {
    const branch = branches[i];
    const path = branch?.[branch.length - 1]?.nodeId ?? `branch_${i}`;
    merged[path] = values[i];
  }
  return merged;
}

// ============================================================================
// Superposition Execution — speculative parallel branch execution
//
// Quantum computing kinda. Run ALL branches of a fork simultaneously,
// let the observation (first-to-complete or external signal) collapse
// the superposition. Vented branches feed the void boundary. The
// complement distribution focuses future speculation on branches most
// likely to survive.
//
// With Lilith at 3μs and compiled execution at 6ns, we can speculatively
// execute entire topology branches in the time one network round trip takes.
// Sequential knots unwind into parallel races. The crossing dissolves
// because both strands run at once.
//
// Recursive: speculative branches can themselves fork speculatively.
// The void boundary prunes the speculation tree from the leaves up --
// branches that consistently lose stop being speculated on (Deceptacon
// applied to execution paths instead of attention heads).
//
// Three phases:
//   SUPERPOSE: fork all branches simultaneously (like quantum superposition)
//   OBSERVE:   first-to-complete or external signal collapses the state
//   VENT:      losers' rejection feeds the void boundary for next round
// ============================================================================

/**
 * Superposition execution context: tracks void boundary across
 * speculative branches for cross-round learning.
 */
export interface SuperpositionContext {
  /** Void boundary: per-branch rejection counts. */
  readonly boundary: VoidBoundary;
  /** Total speculation rounds executed. */
  rounds: number;
  /** Branches pruned by void boundary (consistently lose). */
  pruned: number;
  /** Whether to record branch outcomes in the void boundary. */
  learn: boolean;
}

/**
 * Create a fresh superposition context for a topology with N fork branches.
 */
export function createSuperposition(branchCount: number): SuperpositionContext {
  return {
    boundary: createVoidBoundary(branchCount),
    rounds: 0,
    pruned: 0,
    learn: true,
  };
}

/**
 * Result from superposition execution: includes which branch won.
 */
export interface SuperpositionResult extends CompiledExecutionResult {
  /** Index of the winning branch (for void boundary tracking). */
  winnerBranch: number;
  /** Number of branches that raced. */
  racedBranches: number;
  /** Number of branches pruned by void boundary. */
  prunedBranches: number;
  /** Complement distribution used for this round's speculation. */
  speculationWeights: number[];
}

/**
 * Execute a compiled topology in superposition mode.
 *
 * At each fork, ALL non-pruned branches execute simultaneously.
 * The void boundary's complement distribution determines which branches
 * to speculate on. Branches at floor weight (w=1, maximally rejected)
 * are pruned -- they consistently lose and speculation is wasted.
 *
 * This is the unknotting algorithm applied to execution:
 * - Sequential crossings become parallel races (crossing dissolves)
 * - The void boundary is the knot invariant (tracks which crossings recur)
 * - Each round of speculation is a Reidemeister move (untangles one crossing)
 * - Convergence: when all branches that lose are pruned, the execution
 *   topology is the unknot (zero speculative waste)
 */
export async function executeSuperposition(
  topology: CompiledTopology,
  input: unknown,
  ctx: SuperpositionContext
): Promise<SuperpositionResult> {
  const start = performance.now();
  let payload = input;
  let handlerCalls = 0;
  let winnerBranch = -1;
  let racedBranches = 0;
  let prunedBranches = 0;

  for (const step of topology.steps) {
    if ('kind' in step && step.kind === 'fork') {
      const branches = step.branches;
      const branchCount = branches.length;
      const branchPayloads =
        Array.isArray(payload) && payload.length === branchCount
          ? payload
          : branches.map(() => payload);

      // Compute complement distribution: where the void has accumulated
      // least, the complement peaks -- those are the branches to speculate on.
      //       const dist = complementDistribution(ctx.boundary, 3.0);

      // Determine which branches to speculate on.
      // Branches at floor weight (w=1) are pruned -- they consistently lose.
      const R = ctx.rounds || 1;
      const liveBranches: number[] = [];
      for (let i = 0; i < branchCount; i++) {
        const v = i < ctx.boundary.counts.length ? ctx.boundary.counts[i] : 0;
        const rejectionCount = v ?? 0;
        const w = buleyeanWeight(R, rejectionCount);
        if (w > 1 || ctx.rounds < 3) {
          // Above floor weight, or too early to prune (need warmup data)
          liveBranches.push(i);
        } else {
          prunedBranches++;
        }
      }

      // SUPERPOSE: execute all live branches simultaneously
      if (liveBranches.length === 0) {
        // All pruned -- fall back to branch 0 (the sliver: never zero)
        liveBranches.push(0);
      }

      racedBranches = liveBranches.length;
      const branchPromises = liveBranches.map(async (branchIdx) => {
        let branchPayload = branchPayloads[branchIdx] ?? null;
        const branch = branches[branchIdx];
        if (!branch) return { branchIdx, payload: branchPayload };
        for (const branchStep of branch) {
          branchPayload = await branchStep.handler(
            branchPayload,
            branchStep.properties,
            { nodeId: branchStep.nodeId }
          );
          handlerCalls++;
        }
        return { branchIdx, payload: branchPayload };
      });

      if (step.strategy === 'race') {
        // OBSERVE: first to complete collapses the superposition
        const winner = await Promise.race(branchPromises);
        payload = winner.payload;
        winnerBranch = winner.branchIdx;

        // VENT: record losing branches in void boundary
        if (ctx.learn) {
          for (const idx of liveBranches) {
            if (idx !== winnerBranch) {
              // This branch lost -- record rejection
              if (idx < ctx.boundary.counts.length) {
                updateVoidBoundary(ctx.boundary, idx);
              }
            }
          }
        }
      } else {
        // FOLD: run all, merge results (superposition with full observation)
        const results = await Promise.all(branchPromises);
        payload = mergeCompiledFoldResults(
          results.map((result) => result.payload),
          results.map((result) => branches[result.branchIdx] ?? [])
        );
        const lastResult = results[results.length - 1];
        if (!lastResult) {
          throw new Error('executeSuperposition: fold produced no results');
        }
        winnerBranch = lastResult.branchIdx;
      }

      ctx.rounds++;
      ctx.pruned = prunedBranches;
    } else {
      // Sequential step
      const seqStep = step as CompiledStep;
      payload = await seqStep.handler(payload, seqStep.properties, {
        nodeId: seqStep.nodeId,
      });
      handlerCalls++;
    }
  }

  const executionUs = (performance.now() - start) * 1000;
  const speculationWeights = complementDistribution(ctx.boundary, 3.0);

  return {
    payload,
    executionUs,
    handlerCalls,
    winnerBranch,
    racedBranches,
    prunedBranches,
    speculationWeights,
  };
}

// ============================================================================
// Synchronous Execution — eliminates async/await overhead (~500ns per call)
// ============================================================================

/**
 * Synchronous handler type — no async boundary, no microtask queue.
 * Handlers that return a value directly (not a Promise) can use this
 * path for ~5x faster execution.
 */
export type SyncHandler = (
  payload: unknown,
  properties: Record<string, string>,
  context: { nodeId: string }
) => unknown;

export interface CompiledSyncStep {
  readonly nodeId: string;
  readonly handler: SyncHandler;
  readonly properties: Record<string, string>;
}

export interface CompiledSyncTopology {
  readonly steps: readonly CompiledSyncStep[];
  readonly fingerprint: string;
}

/**
 * Compile a topology into a synchronous execution plan.
 *
 * Only works for linear topologies (no forks). Handlers must be
 * synchronous — if any handler returns a Promise, use executeCompiled.
 */
export function compileSyncTopology(
  ast: GraphAST,
  registry: GnosisRegistry,
  fingerprint: string,
  syncHandlerOverrides?: Map<string, SyncHandler>
): CompiledSyncTopology | null {
  const full = compileTopology(ast, registry, fingerprint);

  // Can't compile forking topologies to sync
  if (full.hasForks) return null;

  const steps: CompiledSyncStep[] = [];
  for (const step of full.steps) {
    if ('kind' in step) return null; // Fork detected

    const override =
      syncHandlerOverrides?.get(step.nodeId) ??
      (step.labels[0] ? syncHandlerOverrides?.get(step.labels[0]) : undefined);
    steps.push({
      nodeId: step.nodeId,
      handler: (override ?? step.handler) as SyncHandler,
      properties: step.properties,
    });
  }

  return { steps, fingerprint };
}

/**
 * Execute a synchronous compiled topology. Zero async overhead.
 *
 * This is the absolute fastest path: no Promises, no microtask queue,
 * no await suspension points. Pure synchronous function calls.
 *
 * For a 3-step topology: ~0.1-0.3μs per execution (3-10M exec/sec).
 */
export function executeCompiledSync(
  topology: CompiledSyncTopology,
  input: unknown
): unknown {
  let payload = input;
  const steps = topology.steps;
  const len = steps.length;

  for (let i = 0; i < len; i++) {
    const step = steps[i];
    if (!step) {
      throw new Error('executeCompiledSync: missing step');
    }
    payload = step.handler(payload, step.properties, { nodeId: step.nodeId });
  }

  return payload;
}

/**
 * Handles the gnosis compile Sync Fork Topology workflow.
 */
export function compileSyncForkTopology(
  ast: GraphAST,
  registry: GnosisRegistry,
  fingerprint: string,
  syncHandlerOverrides: Map<string, SyncHandler>
): CompiledSyncForkTopology | null {
  const full = compileTopology(ast, registry, fingerprint);

  function convertStep(step: CompiledStep): CompiledStep | null {
    const override =
      syncHandlerOverrides.get(step.nodeId) ??
      step.labels.reduce<SyncHandler | undefined>(
        (found, label) => found ?? syncHandlerOverrides.get(label),
        undefined
      );
    if (!override) return null;
    return { ...step, handler: override as GnosisHandler };
  }

  const steps: Array<CompiledStep | CompiledFork> = [];
  for (const step of full.steps) {
    if ('kind' in step) {
      const branches: CompiledStep[][] = [];
      for (const branch of step.branches) {
        const convertedBranch: CompiledStep[] = [];
        for (const branchStep of branch) {
          const converted = convertStep(branchStep);
          if (!converted) return null;
          convertedBranch.push(converted);
        }
        branches.push(convertedBranch);
      }
      steps.push({ ...step, branches });
    } else {
      const converted = convertStep(step);
      if (!converted) return null;
      steps.push(converted);
    }
  }

  return {
    steps,
    nodeCount: full.nodeCount,
    hasForks: full.hasForks,
    fingerprint,
  };
}

/**
 * Handles the gnosis execute Compiled Sync Fork workflow.
 */
export function executeCompiledSyncFork(
  topology: CompiledSyncForkTopology,
  input: unknown
): CompiledExecutionResult {
  const start = performance.now();
  let payload = input;
  let handlerCalls = 0;

  for (const step of topology.steps) {
    if ('kind' in step && step.kind === 'fork') {
      const branchPayloads =
        Array.isArray(payload) && payload.length === step.branches.length
          ? payload
          : step.branches.map(() => payload);

      if (step.strategy === 'race') {
        const branch = step.branches[0] ?? [];
        let branchPayload = branchPayloads[0] ?? null;
        for (const branchStep of branch) {
          branchPayload = (branchStep.handler as unknown as SyncHandler)(
            branchPayload,
            branchStep.properties,
            { nodeId: branchStep.nodeId }
          );
          handlerCalls++;
        }
        payload = branchPayload;
      } else {
        const results: unknown[] = [];
        for (
          let branchIndex = 0;
          branchIndex < step.branches.length;
          branchIndex++
        ) {
          const branch = step.branches[branchIndex] ?? [];
          let branchPayload = branchPayloads[branchIndex] ?? null;
          for (const branchStep of branch) {
            branchPayload = (branchStep.handler as unknown as SyncHandler)(
              branchPayload,
              branchStep.properties,
              { nodeId: branchStep.nodeId }
            );
            handlerCalls++;
          }
          results.push(branchPayload);
        }
        payload = mergeCompiledFoldResults(results, step.branches);
      }
    } else {
      const seqStep = step as CompiledStep;
      payload = (seqStep.handler as unknown as SyncHandler)(
        payload,
        seqStep.properties,
        { nodeId: seqStep.nodeId }
      );
      handlerCalls++;
    }
  }

  return {
    payload,
    executionUs: (performance.now() - start) * 1000,
    handlerCalls,
  };
}

// ============================================================================
// Code Generation — compile topology to a single flat function
// ============================================================================

/**
 * Generate a flat JavaScript function from a compiled topology.
 *
 * The generated function has zero iteration overhead: no for-loop,
 * no array access, no property lookup. Just raw function calls
 * inlined in sequence.
 *
 * Example output for a 3-step topology:
 *   (input, h0, p0, h1, p1, h2, p2) => {
 *     let v = input;
 *     v = h0(v, p0, c0);
 *     v = h1(v, p1, c1);
 *     v = h2(v, p2, c2);
 *     return v;
 *   }
 *
 * The handlers and properties are closed over — no Map lookup,
 * no string comparison, no object allocation in the hot path.
 */
export function codegenSyncExecutor(
  topology: CompiledSyncTopology
): (input: unknown) => unknown {
  const steps = topology.steps;
  const len = steps.length;

  if (len === 0) return (input) => input;

  if (len === 1) {
    const s = steps[0];
    if (!s) return (input) => input;
    const h = s.handler;
    const p = s.properties;
    const c = { nodeId: s.nodeId };
    return (input) => h(input, p, c);
  }

  if (len === 2) {
    const s0 = steps[0],
      s1 = steps[1];
    if (!s0 || !s1) return (input) => input;
    const h0 = s0.handler,
      h1 = s1.handler;
    const p0 = s0.properties,
      p1 = s1.properties;
    const c0 = { nodeId: s0.nodeId },
      c1 = { nodeId: s1.nodeId };
    return (input) => h1(h0(input, p0, c0), p1, c1);
  }

  if (len === 3) {
    const s0 = steps[0],
      s1 = steps[1],
      s2 = steps[2];
    if (!s0 || !s1 || !s2) return (input) => input;
    const h0 = s0.handler,
      h1 = s1.handler,
      h2 = s2.handler;
    const p0 = s0.properties,
      p1 = s1.properties,
      p2 = s2.properties;
    const c0 = { nodeId: s0.nodeId },
      c1 = { nodeId: s1.nodeId },
      c2 = { nodeId: s2.nodeId };
    return (input) => h2(h1(h0(input, p0, c0), p1, c1), p2, c2);
  }

  // General case: pre-extract all handlers/props into locals
  const handlers: SyncHandler[] = new Array(len);
  const props: Record<string, string>[] = new Array(len);
  const contexts: { nodeId: string }[] = new Array(len);
  for (let i = 0; i < len; i++) {
    const step = steps[i];
    if (!step) {
      throw new Error('codegenSyncExecutor: missing step');
    }
    handlers[i] = step.handler;
    props[i] = step.properties;
    contexts[i] = { nodeId: step.nodeId };
  }

  return (input) => {
    let v = input;
    for (let i = 0; i < len; i++) {
      const handler = handlers[i];
      const prop = props[i];
      const context = contexts[i];
      if (!handler || !prop || !context) {
        throw new Error('codegenSyncExecutor: missing cached step');
      }
      v = handler(v, prop, context);
    }
    return v;
  };
}

// ============================================================================
// Cannon Launcher — preloaded, armed, rotating execution surface
// ============================================================================

/**
 * A pre-armed cannon lane holding a compiled topology ready to fire.
 */
interface CannonLane {
  readonly compiled: CompiledTopology;
  armed: boolean;
  executions: number;
  lastExecutionUs: number;
}

/**
 * The Cannon Launcher — preloads compiled topologies and rotates
 * execution through lanes with Wallington cursor advancement.
 *
 * This is the "rubber band" pattern:
 * - All topologies compiled and loaded at init (potential energy)
 * - Cannon cursor rotates through lanes (kinetic energy)
 * - Each execution fires a pre-armed lane (work extraction)
 * - Lane re-arms immediately after firing (elastic rebound)
 */
export class CannonLauncher {
  private lanes: CannonLane[] = [];
  private syncExecutors: Array<(input: unknown) => unknown> = [];
  private cursor = 0;
  private totalExecutions = 0;

  /**
   * Load a compiled topology into a lane.
   * Call this at startup to pre-arm all lanes.
   */
  load(compiled: CompiledTopology): number {
    const laneIndex = this.lanes.length;
    this.lanes.push({
      compiled,
      armed: true,
      executions: 0,
      lastExecutionUs: 0,
    });
    return laneIndex;
  }

  /**
   * Fire the next lane in rotation.
   * Returns the result and advances the cursor (Wallington rotation).
   */
  async fire(
    input: unknown
  ): Promise<CompiledExecutionResult & { laneIndex: number }> {
    if (this.lanes.length === 0) {
      throw new Error('CannonLauncher: no lanes loaded');
    }

    const laneIndex = this.cursor;
    const lane = this.lanes[laneIndex];
    if (!lane) {
      throw new Error('CannonLauncher: missing lane');
    }
    this.cursor = (this.cursor + 1) % this.lanes.length;

    const result = await executeCompiled(lane.compiled, input);

    lane.executions++;
    lane.lastExecutionUs = result.executionUs;
    lane.armed = true; // Re-arm immediately (elastic rebound)
    this.totalExecutions++;

    return { ...result, laneIndex };
  }

  /**
   * Fire the next lane synchronously (zero-async, codegen path).
   * Only works if all lanes were loaded with sync topologies.
   */
  fireSync(input: unknown): unknown {
    if (this.syncExecutors.length === 0) {
      throw new Error(
        'CannonLauncher: no sync executors loaded. Call loadSync() instead of load().'
      );
    }
    const laneIndex = this.cursor;
    this.cursor = (this.cursor + 1) % this.lanes.length;
    const lane = this.lanes[laneIndex];
    const executor = this.syncExecutors[laneIndex];
    if (!lane || !executor) {
      throw new Error('CannonLauncher: missing lane or sync executor');
    }
    const result = executor(input);
    lane.executions++;
    this.totalExecutions++;
    return result;
  }

  /**
   * Load a sync topology into a lane with codegen.
   */
  loadSync(topology: CompiledSyncTopology): number {
    const executor = codegenSyncExecutor(topology);
    const laneIndex = this.lanes.length;
    this.lanes.push({
      compiled: {
        steps: [],
        nodeCount: 0,
        hasForks: false,
        fingerprint: topology.fingerprint,
      },
      armed: true,
      executions: 0,
      lastExecutionUs: 0,
    });
    this.syncExecutors.push(executor);
    return laneIndex;
  }

  /**
   * Fire a specific lane by index (for pinned routing).
   */
  async fireAt(
    laneIndex: number,
    input: unknown
  ): Promise<CompiledExecutionResult> {
    const lane = this.lanes[laneIndex];
    if (!lane) throw new Error(`CannonLauncher: lane ${laneIndex} not found`);

    const result = await executeCompiled(lane.compiled, input);
    lane.executions++;
    lane.lastExecutionUs = result.executionUs;
    return result;
  }

  get laneCount(): number {
    return this.lanes.length;
  }

  get stats(): {
    totalExecutions: number;
    lanes: Array<{ executions: number; lastUs: number; armed: boolean }>;
  } {
    return {
      totalExecutions: this.totalExecutions,
      lanes: this.lanes.map((l) => ({
        executions: l.executions,
        lastUs: l.lastExecutionUs,
        armed: l.armed,
      })),
    };
  }
}

// ─── Compile from Packed Knot ────────────────────────────────────────

/**
 * Compile a topology directly from a packed knot binary.
 * The knot is already toposorted, so steps are in execution order
 * by construction. No AST traversal, no adjacency list, no visited set.
 * Shift and mask. Two bits per step.
 */
export function compileTopologyFromKnot(
  knot: import('./packed-knot.js').PackedKnot,
  registry: GnosisRegistry,
  fingerprint: string
): CompiledTopology {
  const { getCrossing, FORK, RACE, FOLD, VENT } =
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    require('./packed-knot.js') as typeof import('./packed-knot.js');

  const steps: Array<CompiledStep | CompiledFork> = [];
  let i = 0;

  while (i < knot.crossingCount) {
    const op = getCrossing(knot, i);
    const nodeId = knot.nodeIds[i] ?? `knot_${i}`;
    const handler = registry.get(nodeId);

    if (op === FORK) {
      // Collect consecutive fork branches until we hit RACE or FOLD
      const branches: CompiledStep[][] = [];
      const forkStart = i;
      i++;

      // Each subsequent non-FORK crossing before the RACE/FOLD is a branch step
      let currentBranch: CompiledStep[] = [];
      while (i < knot.crossingCount) {
        const nextOp = getCrossing(knot, i);
        const nextNodeId = knot.nodeIds[i] ?? `knot_${i}`;
        const nextHandler = registry.get(nextNodeId);

        if (nextOp === RACE || nextOp === FOLD) {
          if (currentBranch.length > 0) branches.push(currentBranch);
          steps.push({
            kind: 'fork',
            branches:
              branches.length > 0
                ? branches
                : [
                    [
                      {
                        nodeId: knot.nodeIds[forkStart] ?? `knot_${forkStart}`,
                        handler: handler ?? defaultHandler,
                        properties: {},
                        labels: [nodeId],
                      },
                    ],
                  ],
            strategy: nextOp === RACE ? 'race' : 'fold',
          } as CompiledFork);
          i++;
          break;
        }

        currentBranch.push({
          nodeId: nextNodeId,
          handler: nextHandler ?? defaultHandler,
          properties: {},
          labels: [nextNodeId],
        });

        if (nextOp === FORK) {
          // New branch boundary
          branches.push(currentBranch);
          currentBranch = [];
        }
        i++;
      }

      if (currentBranch.length > 0) {
        branches.push(currentBranch);
        // No closing RACE/FOLD found -- emit as fork with fold strategy
        steps.push({
          kind: 'fork',
          branches,
          strategy: 'fold',
        } as CompiledFork);
      }
    } else if (op === VENT) {
      // Vent steps are pass-through (dissipate)
      if (handler) {
        steps.push({
          nodeId,
          handler,
          properties: {},
          labels: [nodeId],
        });
      }
      i++;
    } else {
      // FOLD or RACE as standalone step (no preceding FORK)
      steps.push({
        nodeId,
        handler: handler ?? defaultHandler,
        properties: {},
        labels: [nodeId],
      });
      i++;
    }
  }

  return {
    steps,
    nodeCount: knot.crossingCount,
    hasForks: steps.some(
      (s): s is CompiledFork => 'kind' in s && s.kind === 'fork'
    ),
    fingerprint,
  };
}

const defaultHandler: GnosisHandler = async (payload: unknown) => payload;
