// @a0n/.ignored_gnosis/src/runtime/execution-boundary.ts
/**
 * Lifecycle manager for VoidBoundary during topology execution.
 * Tracks void accumulation through FORK/RACE/FOLD/VENT operations.
 */

import {
  VoidBoundary,
  createVoidBoundary,
  updateVoidBoundary,
  mergeVoidBoundaries,
  buleyeanPositivity,
  buleyeanWeight,
  teleportDeficit,
  complementDistribution,
  shannonEntropy,
  type Walker,
  createWalker,
  c0Update,
  c1Measure,
  c2c3Adapt,
  type Measurement,
  type Gait,
} from '../void';

export interface ExecutionBoundarySnapshot {
  boundary: VoidBoundary;
  positivity: boolean;
  deficit: number;
  entropy: number;
  weights: number[];
  gait: Gait;
  measurement: Measurement | null;
  walkerSteps: number;
}

export class ExecutionBoundary {
  private boundary: VoidBoundary;
  private rounds: number = 0;
  private walker: Walker | null = null;

  constructor(dimensions: number = 0) {
    this.boundary = createVoidBoundary(dimensions);
  }

  /** Called on FORK: create a boundary sized to the number of branches */
  fork(branchCount: number): void {
    if (this.boundary.counts.length === 0) {
      this.boundary = createVoidBoundary(branchCount);
    } else {
      // Expand dimensions if needed
      while (this.boundary.counts.length < branchCount) {
        this.boundary.counts.push(0);
      }
    }
  }

  /** Called on RACE: record a loser in the void boundary */
  raceLoser(loserIdx: number): void {
    if (loserIdx >= 0 && loserIdx < this.boundary.counts.length) {
      updateVoidBoundary(this.boundary, loserIdx);
      this.rounds++;
    }
  }

  /** Called on FOLD: merge another boundary into this one */
  fold(other: VoidBoundary): void {
    this.boundary = mergeVoidBoundaries(this.boundary, other);
    this.rounds++;
  }

  /** Called on VENT: increment void entry for a vented dimension */
  vent(dimensionIdx: number = 0): void {
    if (this.boundary.counts.length === 0) {
      this.boundary = createVoidBoundary(1);
    }
    const idx = Math.min(dimensionIdx, this.boundary.counts.length - 1);
    updateVoidBoundary(this.boundary, Math.max(0, idx));
    this.rounds++;
  }

  /** Get the raw VoidBoundary */
  get void(): VoidBoundary {
    return this.boundary;
  }

  /** Get Buleyean weights for all dimensions */
  get weights(): number[] {
    return this.boundary.counts.map((v) => buleyeanWeight(this.rounds, v));
  }

  /** Check Buleyean positivity */
  get isPositive(): boolean {
    return buleyeanPositivity(this.boundary, this.rounds);
  }

  /** Get deficit */
  get deficit(): number {
    return teleportDeficit(this.boundary);
  }

  /** Create a walker from the boundary if not yet created */
  ensureWalker(): void {
    if (!this.walker) {
      this.walker = createWalker(this.boundary);
    }
  }

  /** Record an execution outcome in the walker's c0 loop */
  recordExecution(nodeIdx: number, rejected: boolean): void {
    if (!this.walker) return;
    const idx = Math.min(nodeIdx, this.walker.boundary.counts.length - 1);
    if (idx < 0) return;
    c0Update(this.walker, idx, rejected ? 1 : 0, rejected ? 0 : 1);
  }

  /** Run c2c3 adaptation on the walker */
  adapt(): void {
    if (!this.walker) return;
    c2c3Adapt(this.walker);
  }

  /** Take a c1 measurement of the walker */
  measurement(): Measurement | null {
    if (!this.walker) return null;
    return c1Measure(this.walker);
  }

  /** Get the walker's current gait */
  gait(): Gait {
    return this.walker ? this.walker.gait : 'stand';
  }

  /** Take a snapshot of the current state */
  snapshot(): ExecutionBoundarySnapshot {
    const dist = complementDistribution(this.boundary);
    return {
      boundary: {
        counts: [...this.boundary.counts],
        totalEntries: this.boundary.totalEntries,
      },
      positivity: this.isPositive,
      deficit: this.deficit,
      entropy: shannonEntropy(dist),
      weights: this.weights,
      gait: this.gait(),
      measurement: this.measurement(),
      walkerSteps: this.walker ? this.walker.steps : 0,
    };
  }

  /** Merge with another ExecutionBoundary */
  mergeWith(other: ExecutionBoundary): void {
    this.fold(other.void);
  }

  /** Clone this boundary */
  clone(): ExecutionBoundary {
    const eb = new ExecutionBoundary(this.boundary.counts.length);
    eb.boundary = {
      counts: [...this.boundary.counts],
      totalEntries: this.boundary.totalEntries,
    };
    eb.rounds = this.rounds;
    if (this.walker) {
      eb.walker = createWalker(
        {
          counts: [...this.walker.boundary.counts],
          totalEntries: this.walker.boundary.totalEntries,
        },
        this.walker.eta,
        this.walker.exploration
      );
      eb.walker.gait = this.walker.gait;
      eb.walker.steps = this.walker.steps;
      eb.walker.score = this.walker.score;
      eb.walker.adaptations = this.walker.adaptations;
      eb.walker.lastAdaptationStep = this.walker.lastAdaptationStep;
    }
    return eb;
  }
}
