// @a0n/.ignored_gnosis/src/runtime/generative-topology.ts
/**
 * Generative Topology -- the spiral
 *
 * A topology executes, observes its own outcomes, and generates a new
 * version of itself. The void boundary is the memory across iterations.
 * The complement distribution is the creative force -- where the void
 * has accumulated least is where to grow new structure.
 *
 * The shape:
 *
 *   execute → observe → void boundary updates →
 *     complement distribution shifts →
 *       generate new topology from the shifted distribution →
 *         execute the NEW topology → observe → ...
 *
 * A lambda that takes a topology and returns a better topology.
 * The fixed point is the unknot. Each iteration is a Reidemeister move.
 * The program rewrites itself by running itself.
 *
 * This is fork/race/fold applied to fork/race/fold:
 *   FORK: generate N variant topologies from complement distribution
 *   RACE: execute all variants, first to complete with best quality wins
 *   FOLD: merge winner's structure into the canonical topology
 *   VENT: record losing variants in the void boundary
 *
 * The Forest convergence loop does this for compilers.
 * Generative topology does this for the programs themselves.
 *
 * Mechanized:
 *   wallington_produces_unknot -- the spiral terminates
 *   wallington_preserves_invariant -- each iteration preserves I/O behavior
 *   reidemeister_preserves_invariant -- each rewrite is safe
 */

import type {
  CompiledTopology,
  CompiledStep,
  CompiledFork,
  CompiledExecutionResult,
  SuperpositionContext,
} from './compiled-topology';

import {
  executeCompiled,
  executeSuperposition,
  createSuperposition,
} from './compiled-topology';

import {
  type VoidBoundary,
  createVoidBoundary,
  updateVoidBoundary,
  complementDistribution,
  sampleComplement,
  teleportDeficit,
  shannonEntropy,
} from '../void';

// ============================================================================
// Types
// ============================================================================

/**
 * A topology variant: a candidate in the generative spiral.
 */
export interface TopologyVariant {
  /** The compiled topology to execute. */
  topology: CompiledTopology;
  /** Which generation created this variant. */
  generation: number;
  /** The mutation that produced this variant from the parent. */
  mutation: TopologyMutation;
  /** Execution result (filled after racing). */
  result?: CompiledExecutionResult;
}

/**
 * A mutation applied to generate a topology variant.
 */
export interface TopologyMutation {
  /** What kind of mutation. */
  kind: 'reorder' | 'prune' | 'parallelize' | 'speculate' | 'identity';
  /** Which step was mutated (index into topology.steps). */
  stepIndex: number;
  /** Human-readable description. */
  description: string;
}

/**
 * The spiral's state: void boundary + generation counter + convergence.
 */
export interface GenerativeSpiralState {
  /** Void boundary tracking which topology variants lose. */
  boundary: VoidBoundary;
  /** Current best topology. */
  bestTopology: CompiledTopology;
  /** Current generation number. */
  generation: number;
  /** History of winning mutations. */
  history: TopologyMutation[];
  /** Whether the spiral has converged (unknot reached). */
  converged: boolean;
  /** Deficit at convergence (0 = unknot). */
  deficit: number;
}

/**
 * Configuration for the generative spiral.
 */
export interface GenerativeSpiralConfig {
  /** Maximum generations before stopping. */
  maxGenerations: number;
  /** Number of variants to generate per generation (fork width). */
  variantsPerGeneration: number;
  /** Convergence threshold: stop when deficit < this. */
  convergenceThreshold: number;
  /** Whether to use superposition execution for variants. */
  useSuperposition: boolean;
  /** Test input to execute topologies against. */
  testInput: unknown;
}

const DEFAULT_SPIRAL_CONFIG: GenerativeSpiralConfig = {
  maxGenerations: 10,
  variantsPerGeneration: 4,
  convergenceThreshold: 0.01,
  useSuperposition: true,
  testInput: null,
};

// ============================================================================
// Generative Spiral
// ============================================================================

/**
 * Run the generative spiral: execute → observe → generate → repeat.
 *
 * Each generation:
 *   1. Generate N variant topologies from the complement distribution
 *   2. Execute all variants (fork/race)
 *   3. Select the winner (fastest, or best quality)
 *   4. Record losers in void boundary (vent)
 *   5. Check convergence (deficit → 0 = unknot)
 *   6. Winner becomes the parent for next generation
 *
 * The spiral terminates when:
 *   - Deficit reaches 0 (unknot -- no further improvement possible)
 *   - Max generations reached
 *   - All variants produce identical results (fixed point)
 */
export async function runGenerativeSpiral(
  initialTopology: CompiledTopology,
  config: Partial<GenerativeSpiralConfig> = {}
): Promise<GenerativeSpiralState> {
  const cfg = { ...DEFAULT_SPIRAL_CONFIG, ...config };

  const state: GenerativeSpiralState = {
    boundary: createVoidBoundary(cfg.variantsPerGeneration),
    bestTopology: initialTopology,
    generation: 0,
    history: [],
    converged: false,
    deficit: Infinity,
  };

  for (let gen = 0; gen < cfg.maxGenerations; gen++) {
    state.generation = gen;

    // ── FORK: generate variant topologies from complement distribution ──
    const variants = generateVariants(
      state.bestTopology,
      state.boundary,
      cfg.variantsPerGeneration,
      gen
    );

    // ── RACE: execute all variants, select winner ──
    const results = await Promise.all(
      variants.map(async (variant) => {
        const result = await executeCompiled(variant.topology, cfg.testInput);
        return { ...variant, result };
      })
    );

    // Select winner: fastest execution time
    let winner = results[0];
    for (let i = 1; i < results.length; i++) {
      if (
        results[i].result &&
        winner.result &&
        results[i].result!.executionUs < winner.result!.executionUs
      ) {
        winner = results[i];
      }
    }

    // ── VENT: record losers in void boundary ──
    for (let i = 0; i < results.length; i++) {
      if (results[i] !== winner && i < state.boundary.counts.length) {
        updateVoidBoundary(state.boundary, i);
      }
    }

    // ── FOLD: winner becomes the new best ──
    state.bestTopology = winner.topology;
    state.history.push(winner.mutation);

    // ── Check convergence ──
    state.deficit = teleportDeficit(state.boundary);
    const dist = complementDistribution(state.boundary);
    const entropy = shannonEntropy(dist);
    const maxEntropy = Math.log(cfg.variantsPerGeneration);

    // Converged when entropy is near maximum (uniform -- all variants equally good)
    // or deficit is near zero (all variants have been tried equally)
    if (
      entropy > maxEntropy * (1 - cfg.convergenceThreshold) ||
      state.deficit === 0
    ) {
      state.converged = true;
      break;
    }
  }

  return state;
}

// ============================================================================
// Variant Generation -- the creative force
// ============================================================================

/**
 * Generate N variant topologies from the complement distribution.
 *
 * Where the void has accumulated LEAST, the complement PEAKS.
 * Those peaks are where to grow new structure -- the unexplored
 * mutations that haven't been rejected yet.
 *
 * Mutation types:
 *   - reorder: swap adjacent steps (changes evaluation order)
 *   - prune: remove a step whose void count is highest (dead path)
 *   - parallelize: convert sequential steps to a fork (add parallelism)
 *   - speculate: add a speculative branch (pre-execute likely next step)
 *   - identity: keep unchanged (the sliver -- always include current best)
 */
function generateVariants(
  parent: CompiledTopology,
  boundary: VoidBoundary,
  count: number,
  generation: number
): TopologyVariant[] {
  const variants: TopologyVariant[] = [];
  const dist = complementDistribution(boundary, 3.0);

  // The sliver: always include the identity (current best, unchanged)
  variants.push({
    topology: parent,
    generation,
    mutation: {
      kind: 'identity',
      stepIndex: -1,
      description: 'Unchanged parent topology (the sliver)',
    },
  });

  // Generate N-1 variants guided by complement distribution
  const mutationKinds: TopologyMutation['kind'][] = [
    'reorder',
    'prune',
    'parallelize',
    'speculate',
  ];

  let rngState = generation * 1664525 + 1013904223;
  const rng = () => {
    rngState = (rngState * 1664525 + 1013904223) & 0x7fffffff;
    return rngState / 0x7fffffff;
  };

  for (let i = 1; i < count; i++) {
    // Sample mutation kind from complement distribution
    const kindIdx = Math.min(
      sampleComplement(dist, rng),
      mutationKinds.length - 1
    );
    const kind = mutationKinds[kindIdx];

    // Choose which step to mutate
    const stepIdx =
      parent.steps.length > 0 ? Math.floor(rng() * parent.steps.length) : 0;

    const mutated = applyMutation(parent, kind, stepIdx);
    variants.push({
      topology: mutated.topology,
      generation,
      mutation: mutated.mutation,
    });
  }

  return variants;
}

/**
 * Apply a mutation to a topology, producing a variant.
 */
function applyMutation(
  parent: CompiledTopology,
  kind: TopologyMutation['kind'],
  stepIndex: number
): { topology: CompiledTopology; mutation: TopologyMutation } {
  const steps = [...parent.steps];

  switch (kind) {
    case 'reorder': {
      // Swap step at stepIndex with step at stepIndex+1
      if (stepIndex < steps.length - 1) {
        const temp = steps[stepIndex];
        steps[stepIndex] = steps[stepIndex + 1];
        steps[stepIndex + 1] = temp;
      }
      return {
        topology: { ...parent, steps },
        mutation: {
          kind,
          stepIndex,
          description: `Reorder: swap steps ${stepIndex} and ${stepIndex + 1}`,
        },
      };
    }

    case 'prune': {
      // Remove the step at stepIndex (if not a fork)
      if (stepIndex < steps.length && !('kind' in steps[stepIndex])) {
        steps.splice(stepIndex, 1);
      }
      return {
        topology: { ...parent, steps, nodeCount: parent.nodeCount - 1 },
        mutation: {
          kind,
          stepIndex,
          description: `Prune: remove step ${stepIndex}`,
        },
      };
    }

    case 'parallelize': {
      // Convert two adjacent sequential steps into a fork
      if (
        stepIndex < steps.length - 1 &&
        !('kind' in steps[stepIndex]) &&
        !('kind' in steps[stepIndex + 1])
      ) {
        const fork: CompiledFork = {
          kind: 'fork',
          branches: [
            [steps[stepIndex] as CompiledStep],
            [steps[stepIndex + 1] as CompiledStep],
          ],
          strategy: 'fold',
        };
        steps.splice(stepIndex, 2, fork);
      }
      return {
        topology: { ...parent, steps, hasForks: true },
        mutation: {
          kind,
          stepIndex,
          description: `Parallelize: fork steps ${stepIndex} and ${
            stepIndex + 1
          }`,
        },
      };
    }

    case 'speculate': {
      // Wrap a sequential step in a race with a speculative copy
      if (stepIndex < steps.length && !('kind' in steps[stepIndex])) {
        const step = steps[stepIndex] as CompiledStep;
        const fork: CompiledFork = {
          kind: 'fork',
          branches: [[step], [step]], // Race two copies
          strategy: 'race',
        };
        steps[stepIndex] = fork;
      }
      return {
        topology: { ...parent, steps, hasForks: true },
        mutation: {
          kind,
          stepIndex,
          description: `Speculate: race step ${stepIndex} against itself`,
        },
      };
    }

    default:
      return {
        topology: parent,
        mutation: {
          kind: 'identity',
          stepIndex: -1,
          description: 'Identity (no mutation)',
        },
      };
  }
}
