// @a0n/.ignored_gnosis/src/runtime/hetero-fabric.ts
import type { StabilityHeteroMoAFabricWitness } from '../betty/stability';
import { QDoc } from '../crdt/qdoc';
import {
  QDocAeonRelay,
  type QDocAeonRelayConfig,
  type QDocAeonRelayStatus,
} from '../crdt/aeon-relay';

const FLOW_HEADER_SIZE = 10;
const DEFAULT_FLOW_STREAM_ID = 1;
const DEFAULT_HEDGE_DELAY_MS = 1;
const DEFAULT_DECAY_HALF_LIFE_MS = 60 * 60 * 1000;

export type HeteroMoAFabricLayerKind = 'cpu' | 'gpu' | 'npu' | 'wasm';

export type HeteroMoAFabricBackendKind =
  | 'cpu'
  | 'webgpu'
  | 'cuda'
  | 'webnn'
  | 'vendor-npu'
  | 'wasm';

export type HeteroMoAFabricScheduleStrategy =
  | 'linear'
  | 'parallel'
  | 'collapse'
  | 'cannon';

export type HeteroMoAFabricLaunchGateState = 'armed' | 'open' | 'closed';

export type HeteroMoAFabricPairRole = 'primary' | 'shadow' | 'adjudicator';

export interface HeteroMoAFabricBackendResult {
  value: unknown;
  bytes?: Uint8Array;
  cpuTimeMs?: number;
  detail?: string;
}

export interface HeteroMoAFabricExecutionContext {
  readonly signal: AbortSignal;
  readonly role: HeteroMoAFabricPairRole;
  readonly lane: number;
  readonly layer: HeteroMoAFabricLayerKind;
  readonly cursorPosition: number;
  readonly frameProtocol: string;
  readonly inputDocument?: QDoc | null;
  readonly communityMemory?: HeteroMoAFabricCommunityMemory | null;
}

export interface HeteroMoAFabricBackendDescriptor {
  readonly id: string;
  readonly label: string;
  readonly kind: HeteroMoAFabricBackendKind;
  readonly layer: HeteroMoAFabricLayerKind;
  readonly predictedLatencyMs?: number;
  readonly priority?: number;
  readonly available?: boolean;
  readonly detail?: string;
  execute(
    input: unknown,
    context: HeteroMoAFabricExecutionContext
  ): Promise<HeteroMoAFabricBackendResult>;
}

export interface HeteroMoAFabricLaneCounts {
  cpu: number;
  gpu: number;
  npu: number;
  wasm: number;
}

export interface HeteroMoAFabricPlan {
  readonly fabricNodeId?: string;
  readonly pairedKernel: string;
  readonly scheduleStrategy: HeteroMoAFabricScheduleStrategy;
  readonly launchGate: HeteroMoAFabricLaunchGateState;
  readonly hedgeDelayMs: number;
  readonly hedgePolicy: string;
  readonly frameProtocol: string;
  readonly laneCounts: HeteroMoAFabricLaneCounts;
}

export interface HeteroMoAFabricRunOptions {
  readonly fabricKey?: string;
  readonly witness?: StabilityHeteroMoAFabricWitness;
  readonly plan?: Partial<HeteroMoAFabricPlan>;
  readonly inputDocument?: QDoc | null;
  readonly communityMemory?: HeteroMoAFabricCommunityMemory | null;
  readonly compareOutputs?: (left: unknown, right: unknown) => boolean;
  readonly isSufficient?: (result: HeteroMoAFabricAttempt) => boolean;
  readonly streamId?: number;
  readonly sequence?: number;
  readonly nowMs?: () => number;
  readonly wallClockMs?: () => number;
}

export interface HeteroMoAFabricAttempt {
  readonly backendId: string;
  readonly backendKind: HeteroMoAFabricBackendKind;
  readonly backendLabel: string;
  readonly layer: HeteroMoAFabricLayerKind;
  readonly role: HeteroMoAFabricPairRole;
  readonly lane: number;
  readonly cursorPosition: number;
  readonly status: 'success' | 'error' | 'aborted' | 'skipped' | 'vented';
  readonly value?: unknown;
  readonly bytes: Uint8Array;
  readonly cpuTimeMs: number;
  readonly wallTimeMs: number;
  readonly startedAtMs: number;
  readonly finishedAtMs: number;
  readonly detail?: string;
}

export interface HeteroMoAFabricPairResult {
  readonly layer: HeteroMoAFabricLayerKind;
  readonly lane: number;
  readonly cursorPosition: number;
  readonly primary: HeteroMoAFabricAttempt;
  readonly shadow: HeteroMoAFabricAttempt;
  readonly accepted: HeteroMoAFabricAttempt | null;
  readonly acceptedAtMs: number | null;
  readonly escalated: boolean;
  readonly agreed: boolean;
  readonly shadowSkipped: boolean;
}

export interface HeteroMoAFabricLayerResult {
  readonly layer: HeteroMoAFabricLayerKind;
  readonly cursorStart: number;
  readonly cursorEnd: number;
  readonly selectedBackends: readonly string[];
  readonly pairs: readonly HeteroMoAFabricPairResult[];
  readonly winner: HeteroMoAFabricAttempt | null;
}

export interface HeteroMoAFabricTelemetry {
  readonly winnerBytes: number;
  readonly loserBytes: number;
  readonly loserCompletions: number;
  readonly ventShare: number;
  readonly skippedHedges: number;
  readonly wallTimeMs: number;
  readonly cpuTimeMs: number;
  readonly queueOccupancy: number;
  readonly cursorPosition: number;
  readonly acceptedAtMs: number;
  /** Projection delay depth: number of layers executed before fold. */
  readonly projectionDelayDepth: number;
  /** 3^D / 2^D advantage ratio from delaying projection across layers. */
  readonly projectionAdvantageRatio: number;
  /** 3^D - 2^D: extra states from operating in trit basis. */
  readonly projectionInformationAdvantage: number;
}

export interface HeteroMoAFabricRunResult {
  readonly fabricKey: string;
  readonly gateOpened: boolean;
  readonly frameProtocol: string;
  readonly winner: HeteroMoAFabricAttempt | null;
  readonly frame: Uint8Array | null;
  readonly layerResults: readonly HeteroMoAFabricLayerResult[];
  readonly telemetry: HeteroMoAFabricTelemetry;
}

export interface HeteroMoAFabricBackendScore {
  score: number;
  wins: number;
  successes: number;
  failures: number;
  disagreements: number;
  skippedHedges: number;
  latencyMeanMs: number;
  latencyJitterMs: number;
  latencySamples: number;
  lastUpdatedAt: number;
}

export interface HeteroMoAFabricRunRecord {
  id: string;
  fabricKey: string;
  acceptedBackendId?: string;
  acceptedLayer?: HeteroMoAFabricLayerKind;
  winnerBytes: number;
  loserBytes: number;
  loserCompletions: number;
  skippedHedges: number;
  ventShare: number;
  wallTimeMs: number;
  cpuTimeMs: number;
  queueOccupancy: number;
  cursorPosition: number;
  createdAt: number;
}

export interface HeteroMoAFabricCommunityMemoryOptions {
  readonly doc?: QDoc;
  readonly relayConfig?: QDocAeonRelayConfig | null;
  readonly namespace?: string;
  readonly decayHalfLifeMs?: number;
}

export interface HeteroMoAFabricPredictedTiming {
  readonly predictedWallTimeMs: number;
  readonly latencyJitterMs: number;
  readonly sampleCount: number;
  readonly arrivalHorizonMs: number;
}

export interface HeteroMoAFabricLaunchScheduleEntry {
  readonly selectionIndex: number;
  readonly backendId: string;
  readonly layer: HeteroMoAFabricLayerKind;
  readonly predictedWallTimeMs: number;
  readonly latencyJitterMs: number;
  readonly sampleCount: number;
  readonly arrivalHorizonMs: number;
  readonly launchOffsetMs: number;
}

type MutableOccupancyCounter = {
  inflight: number;
  peak: number;
};

type RunningAttempt = {
  controller: AbortController;
  startedAtMs: number;
  promise: Promise<HeteroMoAFabricAttempt>;
};

type HeteroMoAFabricSelection = {
  layer: HeteroMoAFabricLayerKind;
  cursorStart: number;
  cursorEnd: number;
  selected: readonly HeteroMoAFabricBackendDescriptor[];
};

type HeteroMoAFabricLaunchTask = {
  selectionIndex: number;
  selection: HeteroMoAFabricSelection;
  backend: HeteroMoAFabricBackendDescriptor;
  layer: HeteroMoAFabricLayerKind;
  lane: number;
  cursorPosition: number;
};

function nowMs(): number {
  return globalThis.performance?.now?.() ?? Date.now();
}

function wallClockMs(): number {
  return Date.now();
}

function bootstrapPredictedLatencyMs(
  descriptor: Pick<
    HeteroMoAFabricBackendDescriptor,
    'kind' | 'layer' | 'predictedLatencyMs'
  >
): number {
  if (descriptor.predictedLatencyMs !== undefined) {
    return Math.max(0, descriptor.predictedLatencyMs);
  }

  switch (descriptor.kind) {
    case 'cuda':
    case 'webgpu':
      return 2;
    case 'webnn':
    case 'vendor-npu':
      return 3;
    case 'cpu':
      return 4;
    case 'wasm':
      return 5;
  }
}

function defaultProcessEnv(): Record<string, string | undefined> {
  if (typeof process === 'undefined') {
    return {};
  }
  return process.env as Record<string, string | undefined>;
}

function clampLaneCount(value: number | undefined, fallback: number): number {
  if (!Number.isFinite(value)) {
    return fallback;
  }
  return Math.max(0, Math.floor(value ?? fallback));
}

function createLaneCounts(
  partial: Partial<HeteroMoAFabricLaneCounts> | undefined
): HeteroMoAFabricLaneCounts {
  if (!partial) {
    return {
      cpu: 1,
      gpu: 0,
      npu: 0,
      wasm: 0,
    };
  }

  const laneCounts: HeteroMoAFabricLaneCounts = {
    cpu: clampLaneCount(partial.cpu, 0),
    gpu: clampLaneCount(partial.gpu, 0),
    npu: clampLaneCount(partial.npu, 0),
    wasm: clampLaneCount(partial.wasm, 0),
  }

  return laneCounts;
}

function stableStringify(value: unknown): string {
  if (value === null) {
    return 'null';
  }
  if (typeof value === 'undefined') {
    return 'undefined';
  }
  if (typeof value === 'string') {
    return JSON.stringify(value);
  }
  if (typeof value === 'number' || typeof value === 'boolean') {
    return JSON.stringify(value);
  }
  if (typeof value === 'bigint') {
    return JSON.stringify(value.toString(10));
  }
  if (value instanceof Uint8Array) {
    return JSON.stringify(Array.from(value));
  }
  if (Array.isArray(value)) {
    return `[${value.map((entry) => stableStringify(entry)).join(',')}]`;
  }
  if (typeof value === 'object') {
    const record = value as Record<string, unknown>;
    return `{${Object.keys(record)
      .sort()
      .map((key) => `${JSON.stringify(key)}:${stableStringify(record[key])}`)
      .join(',')}}`;
  }
  return JSON.stringify(String(value));
}

function toBytes(
  value: unknown,
  inputDocument?: QDoc | null,
  explicitBytes?: Uint8Array
): Uint8Array {
  if (explicitBytes) {
    return explicitBytes;
  }
  if (value instanceof Uint8Array) {
    return value;
  }
  if (ArrayBuffer.isView(value)) {
    const view = value as ArrayBufferView;
    return new Uint8Array(
      view.buffer.slice(view.byteOffset, view.byteOffset + view.byteLength)
    );
  }
  if (value instanceof ArrayBuffer) {
    return new Uint8Array(value.slice(0));
  }
  if (inputDocument) {
    return inputDocument.encodeStateAsUpdate();
  }
  return new TextEncoder().encode(stableStringify(value));
}

function encodeAeon10ByteFrame(
  streamId: number,
  sequence: number,
  flags: number,
  payload: Uint8Array
): Uint8Array {
  const buffer = new Uint8Array(FLOW_HEADER_SIZE + payload.length);
  const view = new DataView(buffer.buffer);
  view.setUint16(0, streamId, false);
  view.setUint32(2, sequence >>> 0, false);
  view.setUint8(6, flags);
  buffer[7] = (payload.length >> 16) & 0xff;
  buffer[8] = (payload.length >> 8) & 0xff;
  buffer[9] = payload.length & 0xff;
  buffer.set(payload, FLOW_HEADER_SIZE);
  return buffer;
}

function decayMultiplier(deltaMs: number, halfLifeMs: number): number {
  if (deltaMs <= 0 || halfLifeMs <= 0) {
    return 1;
  }
  return 2 ** (-deltaMs / halfLifeMs);
}

function applyDecay(
  record: HeteroMoAFabricBackendScore,
  atMs: number,
  halfLifeMs: number
): HeteroMoAFabricBackendScore {
  const factor = decayMultiplier(atMs - record.lastUpdatedAt, halfLifeMs);
  return {
    ...record,
    score: record.score * factor,
    wins: record.wins * factor,
    successes: record.successes * factor,
    failures: record.failures * factor,
    disagreements: record.disagreements * factor,
    skippedHedges: record.skippedHedges * factor,
    latencySamples: record.latencySamples * factor,
    lastUpdatedAt: atMs,
  };
}

function updateLatencyEstimate(
  record: HeteroMoAFabricBackendScore,
  observedWallTimeMs: number
): Pick<
  HeteroMoAFabricBackendScore,
  'latencyMeanMs' | 'latencyJitterMs' | 'latencySamples'
> {
  const sample = Math.max(0, observedWallTimeMs);
  const previousSamples = Math.max(0, record.latencySamples);
  const nextSamples = previousSamples + 1;
  const nextMean =
    previousSamples <= 0
      ? sample
      : (record.latencyMeanMs * previousSamples + sample) / nextSamples;
  const deviation = Math.abs(sample - nextMean);
  const nextJitter =
    previousSamples <= 0
      ? 0
      : (record.latencyJitterMs * previousSamples + deviation) / nextSamples;
  return {
    latencyMeanMs: nextMean,
    latencyJitterMs: nextJitter,
    latencySamples: nextSamples,
  };
}

function attemptSignalStrength(attempt: HeteroMoAFabricAttempt): number {
  switch (attempt.status) {
    case 'success':
      return 5;
    case 'vented':
      return 4;
    case 'error':
      return 3;
    case 'aborted':
      return 2;
    case 'skipped':
      return 1;
  }
}

function chooseRecordedAttempt(
  existing: HeteroMoAFabricAttempt | undefined,
  candidate: HeteroMoAFabricAttempt
): HeteroMoAFabricAttempt {
  if (!existing) {
    return candidate;
  }

  const existingStrength = attemptSignalStrength(existing);
  const candidateStrength = attemptSignalStrength(candidate);
  if (candidateStrength !== existingStrength) {
    return candidateStrength > existingStrength ? candidate : existing;
  }
  if (existing.role !== candidate.role) {
    return candidate.role === 'primary' ? candidate : existing;
  }
  return candidate.finishedAtMs >= existing.finishedAtMs ? candidate : existing;
}

function defaultCompareOutputs(left: unknown, right: unknown): boolean {
  return stableStringify(left) === stableStringify(right);
}

function shouldAcceptSufficientPrimaryBeforeShadowLaunch(
  plan: HeteroMoAFabricPlan
): boolean {
  return (
    plan.hedgePolicy === 'race' || plan.hedgePolicy === 'skip-on-sufficient'
  );
}

function shouldAbortPendingShadowAfterSufficientPrimary(
  plan: HeteroMoAFabricPlan
): boolean {
  return plan.hedgePolicy === 'skip-on-sufficient';
}

function totalLaneCount(plan: HeteroMoAFabricPlan): number {
  return (
    plan.laneCounts.cpu +
    plan.laneCounts.gpu +
    plan.laneCounts.npu +
    plan.laneCounts.wasm
  );
}

function createDefaultPlan(
  partial: Partial<HeteroMoAFabricPlan> | undefined
): HeteroMoAFabricPlan {
  const plan = {
    pairedKernel: partial?.pairedKernel ?? 'mirror',
    scheduleStrategy: partial?.scheduleStrategy ?? 'cannon',
    launchGate: partial?.launchGate ?? 'armed',
    hedgeDelayMs: Math.max(
      0,
      Math.floor(partial?.hedgeDelayMs ?? DEFAULT_HEDGE_DELAY_MS)
    ),
    hedgePolicy: partial?.hedgePolicy ?? 'race',
    frameProtocol: partial?.frameProtocol ?? 'aeon-10-byte-binary',
    laneCounts: createLaneCounts(partial?.laneCounts),
    ...(partial?.fabricNodeId !== undefined
      ? { fabricNodeId: partial.fabricNodeId }
      : {}),
  } satisfies HeteroMoAFabricPlan;
  return plan;
}

function resolvePlan(
  options: HeteroMoAFabricRunOptions
): HeteroMoAFabricPlan {
  const witnessPlan = options.witness
    ? createHeteroMoAFabricPlanFromWitness(options.witness)
    : undefined;

  if (!witnessPlan) {
    return createDefaultPlan(options.plan);
  }

  return createDefaultPlan({
    ...witnessPlan,
    ...options.plan,
    laneCounts: {
      ...witnessPlan.laneCounts,
      ...options.plan?.laneCounts,
    },
  });
}

function laneCountForLayer(
  plan: HeteroMoAFabricPlan,
  layer: HeteroMoAFabricLayerKind
): number {
  switch (layer) {
    case 'cpu':
      return plan.laneCounts.cpu;
    case 'gpu':
      return plan.laneCounts.gpu;
    case 'npu':
      return plan.laneCounts.npu;
    case 'wasm':
      return plan.laneCounts.wasm;
  }
}

function isSuccessfulAttempt(
  attempt: HeteroMoAFabricAttempt
): attempt is HeteroMoAFabricAttempt & { status: 'success'; value: unknown } {
  return attempt.status === 'success';
}

function emptyAttempt(
  backend: HeteroMoAFabricBackendDescriptor,
  role: HeteroMoAFabricPairRole,
  lane: number,
  cursorPosition: number,
  startedAtMs: number,
  finishedAtMs: number,
  detail: string,
  status: HeteroMoAFabricAttempt['status']
): HeteroMoAFabricAttempt {
  return {
    backendId: backend.id,
    backendKind: backend.kind,
    backendLabel: backend.label,
    layer: backend.layer,
    role,
    lane,
    cursorPosition,
    status,
    bytes: new Uint8Array(),
    cpuTimeMs: 0,
    wallTimeMs: Math.max(0, finishedAtMs - startedAtMs),
    startedAtMs,
    finishedAtMs,
    detail,
  };
}

function startBackendAttempt(
  backend: HeteroMoAFabricBackendDescriptor,
  input: unknown,
  context: Omit<HeteroMoAFabricExecutionContext, 'signal'>,
  occupancy: MutableOccupancyCounter,
  clock: () => number,
  externalSignal?: AbortSignal
): RunningAttempt {
  const controller = new AbortController();
  const startedAtMs = clock();
  occupancy.inflight += 1;
  occupancy.peak = Math.max(occupancy.peak, occupancy.inflight);
  const abortFromExternal = () => {
    controller.abort();
  };

  if (externalSignal?.aborted) {
    controller.abort();
  } else {
    externalSignal?.addEventListener('abort', abortFromExternal, {
      once: true,
    });
  }

  return {
    controller,
    startedAtMs,
    promise: (async (): Promise<HeteroMoAFabricAttempt> => {
      let abortListener: (() => void) | null = null;

      try {
        if (controller.signal.aborted) {
          return emptyAttempt(
            backend,
            context.role,
            context.lane,
            context.cursorPosition,
            startedAtMs,
            startedAtMs,
            'attempt aborted before launch',
            'aborted'
          );
        }

        const executionPromise = backend
          .execute(input, {
            ...context,
            signal: controller.signal,
          })
          .then((result) => ({ kind: 'success' as const, result }))
          .catch((error) => ({ kind: 'error' as const, error }));

        const abortPromise = new Promise<{ kind: 'aborted' }>((resolve) => {
          if (controller.signal.aborted) {
            resolve({ kind: 'aborted' });
            return;
          }
          abortListener = () => resolve({ kind: 'aborted' });
          controller.signal.addEventListener('abort', abortListener, {
            once: true,
          });
        });

        const observed = await Promise.race([executionPromise, abortPromise]);
        const finishedAtMs = clock();

        if (observed.kind === 'aborted') {
          return emptyAttempt(
            backend,
            context.role,
            context.lane,
            context.cursorPosition,
            startedAtMs,
            finishedAtMs,
            'attempt aborted after external winner',
            'aborted'
          );
        }

        if (observed.kind === 'error') {
          const detail =
            observed.error instanceof Error
              ? observed.error.message
              : String(observed.error ?? 'unknown');
          return {
            backendId: backend.id,
            backendKind: backend.kind,
            backendLabel: backend.label,
            layer: backend.layer,
            role: context.role,
            lane: context.lane,
            cursorPosition: context.cursorPosition,
            status: controller.signal.aborted ? 'aborted' : 'error',
            bytes: new Uint8Array(),
            cpuTimeMs: 0,
            wallTimeMs: Math.max(0, finishedAtMs - startedAtMs),
            startedAtMs,
            finishedAtMs,
            detail,
          };
        }

        const bytes = toBytes(
          observed.result.value,
          context.inputDocument,
          observed.result.bytes
        );
        const attempt = {
          backendId: backend.id,
          backendKind: backend.kind,
          backendLabel: backend.label,
          layer: backend.layer,
          role: context.role,
          lane: context.lane,
          cursorPosition: context.cursorPosition,
          status: 'success',
          value: observed.result.value,
          bytes,
          cpuTimeMs: Math.max(
            0,
            observed.result.cpuTimeMs ?? finishedAtMs - startedAtMs
          ),
          wallTimeMs: Math.max(0, finishedAtMs - startedAtMs),
          startedAtMs,
          finishedAtMs,
          ...(observed.result.detail !== undefined
            ? { detail: observed.result.detail }
            : {}),
        } satisfies HeteroMoAFabricAttempt;
        return attempt;
      } finally {
        if (abortListener) {
          controller.signal.removeEventListener('abort', abortListener);
        }
        occupancy.inflight = Math.max(0, occupancy.inflight - 1);
        externalSignal?.removeEventListener('abort', abortFromExternal);
      }
    })(),
  };
}

function candidateLayers(
  descriptors: readonly HeteroMoAFabricBackendDescriptor[]
): readonly HeteroMoAFabricLayerKind[] {
  const present = new Set<HeteroMoAFabricLayerKind>();
  for (const descriptor of descriptors) {
    if (descriptor.available === false) {
      continue;
    }
    present.add(descriptor.layer);
  }
  return ['cpu', 'gpu', 'npu', 'wasm'].filter((layer) =>
    present.has(layer as HeteroMoAFabricLayerKind)
  ) as HeteroMoAFabricLayerKind[];
}

function sortBackends(
  descriptors: readonly HeteroMoAFabricBackendDescriptor[]
): HeteroMoAFabricBackendDescriptor[] {
  return [...descriptors].sort((left, right) => {
    const leftPriority = left.priority ?? 0;
    const rightPriority = right.priority ?? 0;
    if (leftPriority !== rightPriority) {
      return rightPriority - leftPriority;
    }
    return left.id.localeCompare(right.id);
  });
}

async function waitForAbortableDelay(
  delayMs: number,
  signal?: AbortSignal
): Promise<boolean> {
  if (delayMs <= 0) {
    return !signal?.aborted;
  }

  if (signal?.aborted) {
    return false;
  }

  return await new Promise<boolean>((resolve) => {
    const onAbort = () => {
      clearTimeout(timer);
      signal?.removeEventListener('abort', onAbort);
      resolve(false);
    };
    const timer = setTimeout(() => {
      signal?.removeEventListener('abort', onAbort);
      resolve(true);
    }, delayMs);

    signal?.addEventListener('abort', onAbort, { once: true });
  });
}

async function runMirroredPair(
  primaryBackend: HeteroMoAFabricBackendDescriptor,
  shadowBackend: HeteroMoAFabricBackendDescriptor,
  input: unknown,
  layer: HeteroMoAFabricLayerKind,
  lane: number,
  cursorPosition: number,
  plan: HeteroMoAFabricPlan,
  options: Required<
    Pick<HeteroMoAFabricRunOptions, 'compareOutputs' | 'isSufficient' | 'nowMs'>
  > &
    Pick<HeteroMoAFabricRunOptions, 'communityMemory' | 'inputDocument'> & {
      abortSignal?: AbortSignal;
    },
  occupancy: MutableOccupancyCounter
): Promise<HeteroMoAFabricPairResult> {
  const pairController = new AbortController();
  const abortPairFromExternal = () => {
    pairController.abort();
  };

  if (options.abortSignal?.aborted) {
    pairController.abort();
  } else {
    options.abortSignal?.addEventListener('abort', abortPairFromExternal, {
      once: true,
    });
  }

  const baseContext = {
    lane,
    layer,
    cursorPosition,
    frameProtocol: plan.frameProtocol,
    ...(options.inputDocument !== undefined
      ? { inputDocument: options.inputDocument }
      : {}),
    ...(options.communityMemory !== undefined
      ? { communityMemory: options.communityMemory }
      : {}),
  } as const;
  const primaryAttempt = startBackendAttempt(
    primaryBackend,
    input,
    {
      ...baseContext,
      role: 'primary',
    },
    occupancy,
    options.nowMs,
    pairController.signal
  );

  try {
    let shadowAttempt: RunningAttempt | null = null;
    let shadowResolved = false;
    const trackShadowAttempt = (attempt: RunningAttempt): RunningAttempt => ({
      ...attempt,
      promise: attempt.promise.then((result) => {
        shadowResolved = true;
        return result;
      }),
    });
    const shadowStartPromise =
      plan.hedgeDelayMs <= 0
        ? Promise.resolve(
            trackShadowAttempt(
              startBackendAttempt(
                shadowBackend,
                input,
                {
                  ...baseContext,
                  role: 'shadow',
                },
                occupancy,
                options.nowMs,
                pairController.signal
              )
            )
          )
        : (async (): Promise<RunningAttempt | null> => {
            const shadowWindowOpened = await waitForAbortableDelay(
              plan.hedgeDelayMs,
              pairController.signal
            );
            if (!shadowWindowOpened) {
              return null;
            }
            shadowAttempt = trackShadowAttempt(
              startBackendAttempt(
                shadowBackend,
                input,
                {
                  ...baseContext,
                  role: 'shadow',
                },
                occupancy,
                options.nowMs,
                pairController.signal
              )
            );
            return shadowAttempt;
          })();

    if (plan.hedgeDelayMs <= 0) {
      shadowAttempt = await shadowStartPromise;
    }

    const primary = await primaryAttempt.promise;
    if (
      isSuccessfulAttempt(primary) &&
      options.isSufficient(primary) &&
      shouldAcceptSufficientPrimaryBeforeShadowLaunch(plan) &&
      shadowAttempt === null
    ) {
      pairController.abort();
      const skipped = emptyAttempt(
        shadowBackend,
        'shadow',
        lane,
        cursorPosition,
        primary.finishedAtMs,
        primary.finishedAtMs,
        'shadow skipped after sufficient primary result',
        'skipped'
      );
      return {
        layer,
        lane,
        cursorPosition,
        primary,
        shadow: skipped,
        accepted: primary,
        acceptedAtMs: primary.finishedAtMs,
        escalated: false,
        agreed: true,
        shadowSkipped: true,
      };
    }

    if (
      isSuccessfulAttempt(primary) &&
      options.isSufficient(primary) &&
      shadowAttempt !== null &&
      !shadowResolved &&
      shouldAbortPendingShadowAfterSufficientPrimary(plan)
    ) {
      pairController.abort();
      const shadow = await shadowAttempt.promise;
      return {
        layer,
        lane,
        cursorPosition,
        primary,
        shadow:
          shadow.status === 'aborted'
            ? shadow
            : {
                ...shadow,
                status: 'aborted',
                detail:
                  shadow.detail ?? 'shadow aborted after sufficient primary result',
              },
        accepted: primary,
        acceptedAtMs: primary.finishedAtMs,
        escalated: false,
        agreed: true,
        shadowSkipped: false,
      };
    }

    const runningShadow = shadowAttempt ?? (await shadowStartPromise);
    if (!runningShadow) {
      pairController.abort();
      const shadow = emptyAttempt(
        shadowBackend,
        'shadow',
        lane,
        cursorPosition,
        primary.finishedAtMs,
        options.nowMs(),
        'shadow aborted before launch after external winner',
        'aborted'
      );
      return {
        layer,
        lane,
        cursorPosition,
        primary,
        shadow,
        accepted: null,
        acceptedAtMs: null,
        escalated: false,
        agreed: false,
        shadowSkipped: false,
      };
    }

    const shadow = await runningShadow.promise;
    const agreed =
      isSuccessfulAttempt(primary) &&
      isSuccessfulAttempt(shadow) &&
      options.compareOutputs(primary.value, shadow.value);
    const accepted =
      agreed && primary.finishedAtMs <= shadow.finishedAtMs
        ? primary
        : agreed
        ? shadow
        : null;
    const ventedPrimary =
      accepted === null
        ? {
            ...primary,
            status: 'vented' as const,
            detail: primary.detail ?? 'vented after mirrored pair disagreement',
          }
        : primary;
    const ventedShadow =
      accepted === null
        ? {
            ...shadow,
            status: 'vented' as const,
            detail: shadow.detail ?? 'vented after mirrored pair disagreement',
          }
        : shadow;

    return {
      layer,
      lane,
      cursorPosition,
      primary: ventedPrimary,
      shadow: ventedShadow,
      accepted,
      acceptedAtMs:
        accepted === null
          ? null
          : Math.max(primary.finishedAtMs, shadow.finishedAtMs),
      escalated: !agreed,
      agreed,
      shadowSkipped: false,
    };
  } finally {
    options.abortSignal?.removeEventListener('abort', abortPairFromExternal);
  }
}

function pairAcceptedAtMs(pair: HeteroMoAFabricPairResult): number | null {
  return pair.acceptedAtMs;
}

function selectAcceptedPair(
  pairs: readonly HeteroMoAFabricPairResult[]
): HeteroMoAFabricPairResult | null {
  return (
    [...pairs]
      .filter((pair) => pair.accepted !== null)
      .sort((left, right) => {
        const leftAcceptedAt = pairAcceptedAtMs(left) ?? Number.POSITIVE_INFINITY;
        const rightAcceptedAt =
          pairAcceptedAtMs(right) ?? Number.POSITIVE_INFINITY;
        if (leftAcceptedAt !== rightAcceptedAt) {
          return leftAcceptedAt - rightAcceptedAt;
        }
        return left.lane - right.lane;
      })[0] ?? null
  );
}

function buildFabricKey(
  descriptors: readonly HeteroMoAFabricBackendDescriptor[],
  options: HeteroMoAFabricRunOptions,
  plan: HeteroMoAFabricPlan
): string {
  if (options.fabricKey) {
    return options.fabricKey;
  }
  if (plan.fabricNodeId) {
    return plan.fabricNodeId;
  }
  return `fabric:${descriptors
    .map((descriptor) => descriptor.id)
    .sort()
    .join('|')}`;
}

async function selectLayerBackends(
  fabricKey: string,
  layer: HeteroMoAFabricLayerKind,
  descriptors: readonly HeteroMoAFabricBackendDescriptor[],
  width: number,
  communityMemory: HeteroMoAFabricCommunityMemory | null | undefined,
  atMs: number
): Promise<{
  cursorStart: number;
  cursorEnd: number;
  selected: readonly HeteroMoAFabricBackendDescriptor[];
}> {
  const available = sortBackends(
    descriptors.filter((descriptor) => descriptor.available !== false)
  );
  if (available.length === 0 || width <= 0) {
    return {
      cursorStart: 0,
      cursorEnd: 0,
      selected: [],
    };
  }

  const ranked = communityMemory
    ? communityMemory.rankBackends(fabricKey, available, atMs)
    : available;
  const cursorStart = communityMemory
    ? communityMemory.currentCursor(fabricKey, layer)
    : 0;
  const selected: HeteroMoAFabricBackendDescriptor[] = [];
  for (let laneIndex = 0; laneIndex < width; laneIndex++) {
    const index = (cursorStart + laneIndex) % ranked.length;
    const choice = ranked[index] ?? ranked[0];
    if (choice === undefined) {
      throw new Error('candidateBackends: missing backend choice');
    }
    selected.push(choice);
  }
  const cursorEnd = (cursorStart + width) % ranked.length;
  if (communityMemory) {
    communityMemory.advanceCursor(fabricKey, layer, cursorEnd, atMs);
  }
  return {
    cursorStart,
    cursorEnd,
    selected,
  };
}

export class HeteroMoAFabricCommunityMemory {
  private readonly doc: QDoc;
  private readonly relay: QDocAeonRelay | null;
  private readonly namespace: string;
  private readonly halfLifeMs: number;
  private readonly backendScores;
  private readonly layerCursors;
  private readonly runLog;

  constructor(options: HeteroMoAFabricCommunityMemoryOptions = {}) {
    this.doc = options.doc ?? new QDoc({ guid: 'hetero-moa-fabric-community' });
    this.relay = options.relayConfig
      ? new QDocAeonRelay(this.doc, options.relayConfig)
      : null;
    this.namespace = options.namespace?.trim() || 'hetero_moa_fabric';
    this.halfLifeMs = Math.max(
      1,
      Math.floor(options.decayHalfLifeMs ?? DEFAULT_DECAY_HALF_LIFE_MS)
    );
    this.backendScores = this.doc.getMap<HeteroMoAFabricBackendScore>(
      `${this.namespace}_backend_scores`
    );
    this.layerCursors = this.doc.getMap<{
      cursor: number;
      lastUpdatedAt: number;
    }>(`${this.namespace}_layer_cursors`);
    this.runLog = this.doc.getArray<HeteroMoAFabricRunRecord>(
      `${this.namespace}_run_log`
    );
  }

  get relayStatus(): QDocAeonRelayStatus | null {
    return this.relay?.status ?? null;
  }

  async connectRelay(): Promise<void> {
    if (!this.relay) {
      return;
    }
    await this.relay.connect();
  }

  disconnectRelay(): void {
    this.relay?.disconnect();
  }

  currentCursor(fabricKey: string, layer: HeteroMoAFabricLayerKind): number {
    const current = this.layerCursors.get(`${fabricKey}:${layer}`);
    return Math.max(0, Math.floor(current?.cursor ?? 0));
  }

  advanceCursor(
    fabricKey: string,
    layer: HeteroMoAFabricLayerKind,
    cursor: number,
    atMs = wallClockMs()
  ): void {
    this.layerCursors.set(`${fabricKey}:${layer}`, {
      cursor: Math.max(0, Math.floor(cursor)),
      lastUpdatedAt: atMs,
    });
  }

  rankBackends(
    fabricKey: string,
    descriptors: readonly HeteroMoAFabricBackendDescriptor[],
    atMs = wallClockMs()
  ): HeteroMoAFabricBackendDescriptor[] {
    return [...descriptors].sort((left, right) => {
      const leftScore = this.readBackendScore(fabricKey, left.id, atMs);
      const rightScore = this.readBackendScore(fabricKey, right.id, atMs);
      if (leftScore.score !== rightScore.score) {
        return rightScore.score - leftScore.score;
      }
      const leftTiming = this.predictBackendTiming(fabricKey, left, atMs);
      const rightTiming = this.predictBackendTiming(fabricKey, right, atMs);
      if (leftTiming.arrivalHorizonMs !== rightTiming.arrivalHorizonMs) {
        return leftTiming.arrivalHorizonMs - rightTiming.arrivalHorizonMs;
      }
      const leftPriority = left.priority ?? 0;
      const rightPriority = right.priority ?? 0;
      if (leftPriority !== rightPriority) {
        return rightPriority - leftPriority;
      }
      return left.id.localeCompare(right.id);
    });
  }

  readBackendScore(
    fabricKey: string,
    backendId: string,
    atMs = wallClockMs()
  ): HeteroMoAFabricBackendScore {
    const raw = this.backendScores.get(`${fabricKey}:${backendId}`);
    if (!raw) {
      return {
        score: 0,
        wins: 0,
        successes: 0,
        failures: 0,
        disagreements: 0,
        skippedHedges: 0,
        latencyMeanMs: 0,
        latencyJitterMs: 0,
        latencySamples: 0,
        lastUpdatedAt: atMs,
      };
    }
    const decayed = applyDecay(raw, atMs, this.halfLifeMs);
    this.backendScores.set(`${fabricKey}:${backendId}`, decayed);
    return decayed;
  }

  predictBackendTiming(
    fabricKey: string,
    descriptor: Pick<
      HeteroMoAFabricBackendDescriptor,
      'id' | 'kind' | 'layer' | 'predictedLatencyMs'
    >,
    atMs = wallClockMs()
  ): HeteroMoAFabricPredictedTiming {
    const score = this.readBackendScore(fabricKey, descriptor.id, atMs);
    const predictedWallTimeMs =
      score.latencySamples > 0
        ? Math.max(0, score.latencyMeanMs)
        : bootstrapPredictedLatencyMs(descriptor);
    const latencyJitterMs =
      score.latencySamples > 0 ? Math.max(0, score.latencyJitterMs) : 0;

    return {
      predictedWallTimeMs,
      latencyJitterMs,
      sampleCount: score.latencySamples,
      arrivalHorizonMs: predictedWallTimeMs + latencyJitterMs,
    };
  }

  createLaunchSchedule(
    fabricKey: string,
    descriptors: readonly Pick<
      HeteroMoAFabricBackendDescriptor,
      'id' | 'kind' | 'layer' | 'predictedLatencyMs'
    >[],
    atMs = wallClockMs()
  ): HeteroMoAFabricLaunchScheduleEntry[] {
    const timings = descriptors.map((descriptor, selectionIndex) => {
      const timing = this.predictBackendTiming(fabricKey, descriptor, atMs);
      return {
        selectionIndex,
        backendId: descriptor.id,
        layer: descriptor.layer,
        ...timing,
      };
    });
    const targetArrivalMs = timings.reduce(
      (max, timing) => Math.max(max, timing.arrivalHorizonMs),
      0
    );

    return timings.map((timing) => ({
      ...timing,
      launchOffsetMs: Math.max(0, targetArrivalMs - timing.arrivalHorizonMs),
    }));
  }

  recordRun(
    fabricKey: string,
    result: HeteroMoAFabricRunResult,
    atMs = wallClockMs()
  ): void {
    const attempted = new Map<string, HeteroMoAFabricAttempt>();
    for (const layerResult of result.layerResults) {
      for (const pair of layerResult.pairs) {
        attempted.set(
          pair.primary.backendId,
          chooseRecordedAttempt(
            attempted.get(pair.primary.backendId),
            pair.primary
          )
        );
        attempted.set(
          pair.shadow.backendId,
          chooseRecordedAttempt(
            attempted.get(pair.shadow.backendId),
            pair.shadow
          )
        );
      }
    }

    for (const [backendId, attempt] of attempted.entries()) {
      const next = applyDecay(
        this.readBackendScore(fabricKey, backendId, atMs),
        atMs,
        this.halfLifeMs
      );
      const won = result.winner?.backendId === backendId ? 1 : 0;
      const succeeded = attempt.status === 'success' ? 1 : 0;
      const failed = attempt.status === 'error' ? 1 : 0;
      const disagreement = result.layerResults.some((layerResult) =>
        layerResult.pairs.some(
          (pair) =>
            pair.escalated &&
            (pair.primary.backendId === backendId ||
              pair.shadow.backendId === backendId)
        )
      )
        ? 1
        : 0;
      const skipped = attempt.status === 'skipped' ? 1 : 0;
      const scoreDelta =
        won * 3 + succeeded * 1 - failed * 2 - disagreement * 2 + skipped * 0.5;
      const latency =
        attempt.status !== 'aborted' && attempt.status !== 'skipped'
          ? updateLatencyEstimate(next, attempt.wallTimeMs)
          : {
              latencyMeanMs: next.latencyMeanMs,
              latencyJitterMs: next.latencyJitterMs,
              latencySamples: next.latencySamples,
            };
      this.backendScores.set(`${fabricKey}:${backendId}`, {
        score: next.score + scoreDelta,
        wins: next.wins + won,
        successes: next.successes + succeeded,
        failures: next.failures + failed,
        disagreements: next.disagreements + disagreement,
        skippedHedges: next.skippedHedges + skipped,
        latencyMeanMs: latency.latencyMeanMs,
        latencyJitterMs: latency.latencyJitterMs,
        latencySamples: latency.latencySamples,
        lastUpdatedAt: atMs,
      });
    }

    const runRecord: HeteroMoAFabricRunRecord = {
      id: `${fabricKey}:${atMs}`,
      fabricKey,
      winnerBytes: result.telemetry.winnerBytes,
      loserBytes: result.telemetry.loserBytes,
      loserCompletions: result.telemetry.loserCompletions,
      skippedHedges: result.telemetry.skippedHedges,
      ventShare: result.telemetry.ventShare,
      wallTimeMs: result.telemetry.wallTimeMs,
      cpuTimeMs: result.telemetry.cpuTimeMs,
      queueOccupancy: result.telemetry.queueOccupancy,
      cursorPosition: result.telemetry.cursorPosition,
      createdAt: atMs,
    };
    if (result.winner?.backendId !== undefined) {
      runRecord.acceptedBackendId = result.winner.backendId;
    }
    if (result.winner?.layer !== undefined) {
      runRecord.acceptedLayer = result.winner.layer;
    }
    this.runLog.push([runRecord]);
  }

  runs(): HeteroMoAFabricRunRecord[] {
    return this.runLog.toArray();
  }

  snapshotScores(
    fabricKey: string,
    atMs = wallClockMs()
  ): Record<string, HeteroMoAFabricBackendScore> {
    const snapshot: Record<string, HeteroMoAFabricBackendScore> = {};
    for (const [key] of this.backendScores.entries()) {
      if (!key.startsWith(`${fabricKey}:`)) {
        continue;
      }
      snapshot[key.slice(fabricKey.length + 1)] = this.readBackendScore(
        fabricKey,
        key.slice(fabricKey.length + 1),
        atMs
      );
    }
    return snapshot;
  }
}

/**
 * Creates the Hetero Mo AFabric Plan From Witness.
 */
export function createHeteroMoAFabricPlanFromWitness(
  witness: StabilityHeteroMoAFabricWitness
): HeteroMoAFabricPlan {
  return {
    fabricNodeId: witness.fabricNodeId,
    pairedKernel: witness.pairedKernel,
    scheduleStrategy:
      witness.scheduleStrategy as HeteroMoAFabricScheduleStrategy,
    launchGate: witness.launchGate as HeteroMoAFabricLaunchGateState,
    hedgeDelayMs: Math.max(0, witness.hedgeDelayTicks),
    hedgePolicy: witness.hedgePolicy,
    frameProtocol: witness.frameProtocol,
    laneCounts: {
      cpu: witness.cpuLaneCount,
      gpu: witness.gpuLaneCount,
      npu: witness.npuLaneCount,
      wasm: witness.wasmLaneCount,
    },
  };
}

/**
 * Creates the Dash Relay Config From Env.
 */
export function createDashRelayConfigFromEnv(
  env: Record<string, string | undefined> = defaultProcessEnv()
): QDocAeonRelayConfig | null {
  const url =
    env.GNOSIS_DASHRELAY_URL ??
    env.DASHRELAY_URL ??
    env.GNOSIS_AEON_RELAY_URL ??
    env.AEON_RELAY_URL;
  const roomName =
    env.GNOSIS_DASHRELAY_ROOM ??
    env.DASHRELAY_ROOM ??
    env.GNOSIS_AEON_RELAY_ROOM ??
    env.AEON_RELAY_ROOM;

  if (!url || !roomName) {
    return null;
  }

  const config: QDocAeonRelayConfig = {
    url,
    roomName,
    protocol:
      env.GNOSIS_DASHRELAY_PROTOCOL ??
      env.DASHRELAY_PROTOCOL ??
      env.GNOSIS_AEON_RELAY_PROTOCOL ??
      env.AEON_RELAY_PROTOCOL ??
      'dashrelay-v1',
    joinMode:
      env.GNOSIS_DASHRELAY_MODE ??
      env.DASHRELAY_MODE ??
      env.GNOSIS_AEON_RELAY_MODE ??
      env.AEON_RELAY_MODE ??
      'qdoc-authoritative',
    relayProduct:
      env.GNOSIS_DASHRELAY_PRODUCT ??
      env.DASHRELAY_PRODUCT ??
      env.GNOSIS_AEON_RELAY_PRODUCT ??
      env.AEON_RELAY_PRODUCT ??
      'aeon-relay',
    readyStrategy:
      env.GNOSIS_DASHRELAY_READY_STRATEGY === 'joined' ? 'joined' : 'snapshot',
  };
  const apiKey =
    env.GNOSIS_DASHRELAY_KEY ??
    env.DASHRELAY_API_KEY ??
    env.GNOSIS_AEON_RELAY_KEY ??
    env.AEON_RELAY_KEY;
  if (apiKey !== undefined) {
    config.apiKey = apiKey;
  }
  const clientId =
    env.GNOSIS_DASHRELAY_CLIENT_ID ??
    env.DASHRELAY_CLIENT_ID ??
    env.GNOSIS_AEON_RELAY_CLIENT_ID ??
    env.AEON_RELAY_CLIENT_ID;
  if (clientId !== undefined) {
    config.clientId = clientId;
  }
  return config;
}

function parseCommandSpec(raw: string): string[] {
  const trimmed = raw.trim();
  if (trimmed.length === 0) {
    throw new Error('Runner command cannot be empty');
  }
  if (trimmed.startsWith('[')) {
    const parsed = JSON.parse(trimmed) as unknown;
    if (
      !Array.isArray(parsed) ||
      parsed.length === 0 ||
      !parsed.every(
        (entry) => typeof entry === 'string' && entry.trim().length > 0
      )
    ) {
      throw new Error('Runner JSON command must be a non-empty string array');
    }
    return parsed;
  }
  return trimmed.split(/\s+/);
}

export interface CommandFabricBackendOptions {
  readonly id: string;
  readonly label: string;
  readonly kind: Exclude<
    HeteroMoAFabricBackendKind,
    'cpu' | 'webgpu' | 'webnn'
  >;
  readonly layer: HeteroMoAFabricLayerKind;
  readonly command: readonly string[] | string;
  readonly priority?: number;
  readonly detail?: string;
}

/**
 * Creates the Command Fabric Backend.
 */
export function createCommandFabricBackend(
  options: CommandFabricBackendOptions
): HeteroMoAFabricBackendDescriptor {
  const command =
    typeof options.command === 'string'
      ? parseCommandSpec(options.command)
      : [...options.command];

  if (command.length === 0) {
    throw new Error(
      `Command backend '${options.id}' requires a non-empty command`
    );
  }

  const descriptor: {
    id: string;
    label: string;
    kind: HeteroMoAFabricBackendDescriptor['kind'];
    layer: HeteroMoAFabricLayerKind;
    priority?: number;
    detail?: string;
    execute(
      input: unknown,
      context: HeteroMoAFabricExecutionContext
    ): Promise<HeteroMoAFabricBackendResult>;
  } = {
    id: options.id,
    label: options.label,
    kind: options.kind,
    layer: options.layer,
    async execute(input, context) {
      const { spawn } = await import('node:child_process');
      const commandName = command[0];
      if (commandName === undefined) {
        throw new Error(
          `Command backend '${options.id}' requires a first command segment`
        );
      }
      const child = spawn(commandName, command.slice(1), {
        stdio: ['pipe', 'pipe', 'pipe'],
        env: process.env,
      }) as import('node:child_process').ChildProcessWithoutNullStreams;
      const payload = JSON.stringify({
        input: stableStringify(input),
        frameProtocol: context.frameProtocol,
        lane: context.lane,
        role: context.role,
        layer: context.layer,
      });
      child.stdin.write(payload);
      child.stdin.end();

      context.signal.addEventListener(
        'abort',
        () => {
          child.kill();
        },
        { once: true }
      );

      const stdoutChunks: Buffer[] = [];
      const stderrChunks: Buffer[] = [];
      child.stdout.on('data', (chunk: Buffer) => stdoutChunks.push(Buffer.from(chunk)));
      child.stderr.on('data', (chunk: Buffer) => stderrChunks.push(Buffer.from(chunk)));
      const [exitCode] = await new Promise<
        [number | null, NodeJS.Signals | null]
      >((resolve) => {
        child.once(
          'close',
          (code: number | null, signal: NodeJS.Signals | null) =>
            resolve([code, signal])
        );
      });

      if (exitCode !== 0) {
        throw new Error(
          Buffer.concat(stderrChunks).toString('utf8') ||
            `${options.id} exited with code ${exitCode}`
        );
      }

      const stdout = Buffer.concat(stdoutChunks).toString('utf8').trim();
      if (stdout.length === 0) {
        const result: HeteroMoAFabricBackendResult = {
          value: null,
          bytes: new Uint8Array(),
          detail: `${options.id} returned an empty payload`,
        };
        return result;
      }

      const parsed = JSON.parse(stdout) as {
        value?: unknown;
        bytesBase64?: string;
        cpuTimeMs?: number;
        detail?: string;
      };

      const result: HeteroMoAFabricBackendResult = {
        value: parsed.value,
      };
      if (parsed.bytesBase64 !== undefined) {
        result.bytes = Uint8Array.from(
          Buffer.from(parsed.bytesBase64, 'base64')
        );
      }
      if (parsed.cpuTimeMs !== undefined) {
        result.cpuTimeMs = parsed.cpuTimeMs;
      }
      if (parsed.detail !== undefined) {
        result.detail = parsed.detail;
      }
      return result;
    },
  };
  if (options.priority !== undefined) {
    descriptor.priority = options.priority;
  }
  if (options.detail !== undefined) {
    descriptor.detail = options.detail;
  }
  return descriptor;
}

/**
 * Creates the Command Fabric Backends From Env.
 */
export function createCommandFabricBackendsFromEnv(
  env: Record<string, string | undefined> = defaultProcessEnv()
): HeteroMoAFabricBackendDescriptor[] {
  const backends: HeteroMoAFabricBackendDescriptor[] = [];
  const runnerSpecs = [
    {
      value: env.GNOSIS_CUDA_RUNNER,
      options: {
        id: 'cuda-env',
        label: 'CUDA runner',
        kind: 'cuda' as const,
        layer: 'gpu' as const,
        detail: 'Configured through GNOSIS_CUDA_RUNNER',
      },
    },
    {
      value: env.GNOSIS_VENDOR_NPU_RUNNER,
      options: {
        id: 'vendor-npu-env',
        label: 'Vendor NPU runner',
        kind: 'vendor-npu' as const,
        layer: 'npu' as const,
        detail: 'Configured through GNOSIS_VENDOR_NPU_RUNNER',
      },
    },
    {
      value: env.GNOSIS_WASM_RUNNER,
      options: {
        id: 'wasm-env',
        label: 'WASM runner',
        kind: 'wasm' as const,
        layer: 'wasm' as const,
        detail: 'Configured through GNOSIS_WASM_RUNNER',
      },
    },
  ] as const;

  for (const runner of runnerSpecs) {
    if (!runner.value) {
      continue;
    }
    backends.push(
      createCommandFabricBackend({
        ...runner.options,
        command: runner.value,
      })
    );
  }

  return backends;
}

function createLaunchScheduleEntries(
  fabricKey: string,
  tasks: readonly HeteroMoAFabricLaunchTask[],
  communityMemory: HeteroMoAFabricCommunityMemory | null | undefined,
  atMs: number
): HeteroMoAFabricLaunchScheduleEntry[] {
  if (communityMemory) {
    return communityMemory.createLaunchSchedule(
      fabricKey,
      tasks.map((task) => task.backend),
      atMs
    );
  }

  const predictions = tasks.map((task) => ({
    selectionIndex: task.selectionIndex,
    backendId: task.backend.id,
    layer: task.layer,
    predictedWallTimeMs: bootstrapPredictedLatencyMs(task.backend),
    latencyJitterMs: 0,
    sampleCount: 0,
  }));
  const targetArrivalMs = predictions.reduce(
    (max, prediction) => Math.max(max, prediction.predictedWallTimeMs),
    0
  );

  return predictions.map((prediction) => ({
    ...prediction,
    arrivalHorizonMs:
      prediction.predictedWallTimeMs + prediction.latencyJitterMs,
    launchOffsetMs: Math.max(
      0,
      targetArrivalMs -
        (prediction.predictedWallTimeMs + prediction.latencyJitterMs)
    ),
  }));
}

async function launchMetaLayerTask(
  task: HeteroMoAFabricLaunchTask,
  launchOffsetMs: number,
  input: unknown,
  plan: HeteroMoAFabricPlan,
  options: Required<
    Pick<HeteroMoAFabricRunOptions, 'compareOutputs' | 'isSufficient' | 'nowMs'>
  > &
    Pick<HeteroMoAFabricRunOptions, 'communityMemory' | 'inputDocument'>,
  occupancy: MutableOccupancyCounter,
  abortSignal: AbortSignal
): Promise<{
  layer: HeteroMoAFabricLayerKind;
  pair: HeteroMoAFabricPairResult | null;
}> {
  const launchReady = await waitForAbortableDelay(launchOffsetMs, abortSignal);
  if (!launchReady) {
    return {
      layer: task.layer,
      pair: null,
    };
  }

  const shadowBackend =
    task.selection.selected.length > 1
      ? task.selection.selected[
          (task.lane + 1) % task.selection.selected.length
        ] ??
        task.selection.selected[0] ??
        task.backend
      : task.backend;

  return {
    layer: task.layer,
    pair: await runMirroredPair(
      task.backend,
      shadowBackend,
      input,
      task.layer,
      task.lane,
      task.cursorPosition,
      plan,
      {
        ...options,
        abortSignal,
      },
      occupancy
    ),
  };
}

/**
 * Handles the gnosis run Hetero Mo AFabric workflow.
 */
export async function runHeteroMoAFabric(
  descriptors: readonly HeteroMoAFabricBackendDescriptor[],
  input: unknown,
  options: HeteroMoAFabricRunOptions = {}
): Promise<HeteroMoAFabricRunResult> {
  const plan = resolvePlan(options);
  const activeDescriptors = descriptors.filter(
    (descriptor) => descriptor.available !== false
  );
  const fabricKey = buildFabricKey(activeDescriptors, options, plan);
  const layers = candidateLayers(activeDescriptors);
  const compareOutputs = options.compareOutputs ?? defaultCompareOutputs;
  const sufficient =
    options.isSufficient ??
    ((attempt: HeteroMoAFabricAttempt): boolean =>
      isSuccessfulAttempt(attempt));
  const clock = options.nowMs ?? nowMs;
  const wallClock = options.wallClockMs ?? wallClockMs;
  const occupancy: MutableOccupancyCounter = { inflight: 0, peak: 0 };
  const runStartedAt = clock();
  const createEmptyRunResult = (): HeteroMoAFabricRunResult => ({
    fabricKey,
    gateOpened: false,
    frameProtocol: plan.frameProtocol,
    winner: null,
    frame: null,
    layerResults: [],
    telemetry: {
      winnerBytes: 0,
      loserBytes: 0,
      loserCompletions: 0,
      ventShare: 0,
      skippedHedges: 0,
      wallTimeMs: 0,
      cpuTimeMs: 0,
      queueOccupancy: 0,
      cursorPosition: 0,
      acceptedAtMs: runStartedAt,
      projectionDelayDepth: 0,
      projectionAdvantageRatio: 1,
      projectionInformationAdvantage: 0,
    },
  });

  if (
    plan.launchGate === 'closed' ||
    activeDescriptors.length === 0 ||
    totalLaneCount(plan) <= 0
  ) {
    const emptyResult = createEmptyRunResult();
    options.communityMemory?.recordRun(fabricKey, emptyResult, wallClock());
    return emptyResult;
  }

  const layerResults: HeteroMoAFabricLayerResult[] = [];
  let gateOpened = true;
  const selections: HeteroMoAFabricSelection[] = [];

  for (const layer of layers) {
    const width = laneCountForLayer(plan, layer);
    const layerDescriptors = activeDescriptors.filter(
      (descriptor) => descriptor.layer === layer
    );
    const selection = await selectLayerBackends(
      fabricKey,
      layer,
      layerDescriptors,
      width,
      options.communityMemory,
      wallClock()
    );
    selections.push({
      layer,
      ...selection,
    });
  }

  if (!selections.some((selection) => selection.selected.length > 0)) {
    const emptyResult = createEmptyRunResult();
    options.communityMemory?.recordRun(fabricKey, emptyResult, wallClock());
    return emptyResult;
  }

  if (plan.scheduleStrategy === 'linear') {
    for (const selection of selections) {
      const pairs: HeteroMoAFabricPairResult[] = [];

      for (let lane = 0; lane < selection.selected.length; lane++) {
        const backend = selection.selected[lane];
        if (!backend) {
          continue;
        }
        const shadowBackend =
          selection.selected.length > 1
            ? selection.selected[(lane + 1) % selection.selected.length] ??
              backend
            : backend;
        const pair = await runMirroredPair(
          backend,
          shadowBackend,
          input,
          selection.layer,
          lane,
          selection.cursorStart + lane,
          plan,
          {
            compareOutputs,
            isSufficient: sufficient,
            nowMs: clock,
            ...(options.communityMemory !== undefined
              ? { communityMemory: options.communityMemory }
              : {}),
            ...(options.inputDocument !== undefined
              ? { inputDocument: options.inputDocument }
              : {}),
          },
          occupancy
        );
        pairs.push(pair);
        if (pair.accepted) {
          break;
        }
      }

      const acceptedPair = selectAcceptedPair(pairs);

      layerResults.push({
        layer: selection.layer,
        cursorStart: selection.cursorStart,
        cursorEnd: selection.cursorEnd,
        selectedBackends: selection.selected.map((descriptor) => descriptor.id),
        pairs,
        winner: acceptedPair?.accepted ?? null,
      });
    }
  } else {
    const launchTasks: HeteroMoAFabricLaunchTask[] = [];
    let selectionIndex = 0;
    for (const selection of selections) {
      for (let lane = 0; lane < selection.selected.length; lane++) {
        const backend = selection.selected[lane];
        if (!backend) {
          continue;
        }
        launchTasks.push({
          selectionIndex,
          selection,
          backend,
          layer: selection.layer,
          lane,
          cursorPosition: selection.cursorStart + lane,
        });
        selectionIndex += 1;
      }
    }

    gateOpened = gateOpened && launchTasks.length > 0;
    const schedule = createLaunchScheduleEntries(
      fabricKey,
      launchTasks,
      options.communityMemory,
      wallClock()
    );
    const launchOffsetsBySelectionIndex = new Map(
      schedule.map((entry) => [entry.selectionIndex, entry.launchOffsetMs])
    );
    const metaRaceController = new AbortController();
      const taskPromises = launchTasks.map((task) =>
        launchMetaLayerTask(
          task,
          launchOffsetsBySelectionIndex.get(task.selectionIndex) ?? 0,
          input,
          plan,
          {
            compareOutputs,
            isSufficient: sufficient,
            nowMs: clock,
            ...(options.communityMemory !== undefined
              ? { communityMemory: options.communityMemory }
              : {}),
            ...(options.inputDocument !== undefined
              ? { inputDocument: options.inputDocument }
              : {}),
          },
          occupancy,
          metaRaceController.signal
        )
      );

    try {
      await Promise.any(
        taskPromises.map((promise) =>
          promise.then((result) => {
            if (result.pair?.accepted) {
              return result.pair.accepted;
            }
            return Promise.reject(new Error('pair did not accept'));
          })
        )
      );
      metaRaceController.abort();
    } catch {
      // No accepting pair completed; allow all tasks to settle normally.
    }

    const settledPairs = await Promise.all(taskPromises);
    const pairsByLayer = new Map<
      HeteroMoAFabricLayerKind,
      HeteroMoAFabricPairResult[]
    >();
    for (const settled of settledPairs) {
      if (!settled.pair) {
        continue;
      }
      const bucket = pairsByLayer.get(settled.layer) ?? [];
      bucket.push(settled.pair);
      pairsByLayer.set(settled.layer, bucket);
    }

    for (const selection of selections) {
      const pairs = (pairsByLayer.get(selection.layer) ?? []).sort(
        (left, right) => left.lane - right.lane
      );
      const acceptedPair = selectAcceptedPair(pairs);

      layerResults.push({
        layer: selection.layer,
        cursorStart: selection.cursorStart,
        cursorEnd: selection.cursorEnd,
        selectedBackends: selection.selected.map((descriptor) => descriptor.id),
        pairs,
        winner: acceptedPair?.accepted ?? null,
      });
    }
  }

  const acceptedPairs = layerResults
    .flatMap((layerResult) => layerResult.pairs)
    .filter((pair) => pair.accepted !== null)
    .sort((left, right) => {
      const leftAcceptedAt = pairAcceptedAtMs(left) ?? Number.POSITIVE_INFINITY;
      const rightAcceptedAt =
        pairAcceptedAtMs(right) ?? Number.POSITIVE_INFINITY;
      if (leftAcceptedAt !== rightAcceptedAt) {
        return leftAcceptedAt - rightAcceptedAt;
      }
      return left.lane - right.lane;
    });
  const acceptedPair = acceptedPairs[0] ?? null;
  const winner = acceptedPair?.accepted ?? null;
  const acceptedAtMs = acceptedPair
    ? (pairAcceptedAtMs(acceptedPair) ?? clock())
    : clock();
  const attempts = layerResults.flatMap((layerResult) =>
    layerResult.pairs.flatMap((pair) => [pair.primary, pair.shadow])
  );
  const skippedHedges = attempts.filter(
    (attempt) => attempt.status === 'skipped'
  ).length;
  const loserAttempts = attempts.filter(
    (attempt) => attempt.status === 'success' && attempt !== winner
  );
  const loserCompletions = loserAttempts.length;
  const loserBytes = loserAttempts.reduce(
    (sum, attempt) => sum + attempt.bytes.length,
    0
  );
  const winnerBytes = winner?.bytes.length ?? 0;
  const ventedAttempts = attempts.filter(
    (attempt) =>
      attempt.status === 'error' ||
      attempt.status === 'vented' ||
      attempt.status === 'aborted'
  ).length;
  const cpuTimeMs = attempts.reduce(
    (sum, attempt) => sum + attempt.cpuTimeMs,
    0
  );
  const cursorPosition =
    layerResults.length === 0
      ? 0
      : Math.max(...layerResults.map((result) => result.cursorEnd));
  const payload = winner?.bytes ?? new Uint8Array();
  const FLOW_FLAG_KNOT = 0b01000000;
  const frame =
    winner && plan.frameProtocol === 'aeon-knot-binary'
      ? encodeAeon10ByteFrame(
          options.streamId ?? DEFAULT_FLOW_STREAM_ID,
          options.sequence ?? 1,
          FLOW_FLAG_KNOT,
          payload
        )
      : winner && plan.frameProtocol === 'aeon-10-byte-binary'
      ? encodeAeon10ByteFrame(
          options.streamId ?? DEFAULT_FLOW_STREAM_ID,
          options.sequence ?? 1,
          0,
          payload
        )
      : winner
      ? payload
      : null;

  const result: HeteroMoAFabricRunResult = {
    fabricKey,
    gateOpened,
    frameProtocol: plan.frameProtocol,
    winner,
    frame,
    layerResults,
    telemetry: {
      winnerBytes,
      loserBytes,
      loserCompletions,
      ventShare: attempts.length === 0 ? 0 : ventedAttempts / attempts.length,
      skippedHedges,
      wallTimeMs: Math.max(0, acceptedAtMs - runStartedAt),
      cpuTimeMs,
      queueOccupancy: occupancy.peak,
      cursorPosition,
      acceptedAtMs,
      // Projection delay: each layer is a deferred projection step.
      // The fabric races across CPU/GPU/NPU/WASM without folding
      // until the winner is selected — a Wallington rotation.
      projectionDelayDepth: layerResults.length,
      projectionAdvantageRatio: layerResults.length === 0 ? 1 : 3 ** layerResults.length / 2 ** layerResults.length,
      projectionInformationAdvantage: layerResults.length === 0 ? 0 : 3 ** layerResults.length - 2 ** layerResults.length,
    },
  };

  options.communityMemory?.recordRun(fabricKey, result, wallClock());
  return result;
}
