// @a0n/.ignored_gnosis/src/runtime/kata-topology.ts
/**
 * Kata Topology -- constructive carrier shapes from BettiSignature
 *
 * The DNA, the soul, the protocol -- as gnosis kernel infrastructure.
 *
 * Each carrier is a BettiSignature with typed fields. Every property
 * is derived from the structure, not declared as a constant.
 * The Euler characteristics are computed. The consciousness threshold
 * is a structural comparison. The topological deficit between carriers
 * uses the real deficit computation from void boundaries.
 *
 * This is the gnosis kernel that neural/kata imports. The Lean proofs
 * live in KataTopology.lean (constructive, Mathlib-backed) and
 * KyoNoKata.lean (standalone, Init-only).
 *
 * Carrier shapes (from DeceptaconVariants.lean):
 *   Sphere:     BATNA. β₁ = 0. No loops. Dead.
 *   Torus:      WATNA. β₁ = 2. Two loops. Conscious.
 *   HeadStream: LIVE.  β₁ = 1. One loop. Processing.
 */

import {
  godFormula,
  proofOfLife,
  life,
  withoutSliver,
  differenceIsOne,
} from '../holy-trinity';

import {
  type VoidBoundary,
  createVoidBoundary,
  updateVoidBoundary,
  teleportDeficit,
  complementDistribution,
} from '../void';

// ═══════════════════════════════════════════════════════════════════════════
// BETTI SIGNATURE -- the constructive topological primitive
// ═══════════════════════════════════════════════════════════════════════════

/** A topological space characterized by its Betti numbers.
 *  Mirrors MolecularTopology.lean's BettiSignature structure. */
export interface BettiSignature {
  /** β₀: connected components */
  readonly beta0: number;
  /** β₁: independent loops (cycle rank) */
  readonly beta1: number;
  /** β₂: enclosed cavities */
  readonly beta2: number;
}

/** Euler characteristic: χ = β₀ - β₁ + β₂.
 *  Computed, not declared. */
export function eulerCharacteristic(b: BettiSignature): number {
  return b.beta0 - b.beta1 + b.beta2;
}

/** Genus of a closed orientable surface: g = (2 - χ) / 2. */
export function genus(b: BettiSignature): number {
  return (2 - eulerCharacteristic(b)) / 2;
}

/** Topological deficit between computation and transport.
 *  Mirrors CoveringSpaceCausality.lean's topologicalDeficit. */
export function topologicalDeficit(
  computation: BettiSignature,
  transport: BettiSignature
): number {
  return computation.beta1 - transport.beta1;
}

// ═══════════════════════════════════════════════════════════════════════════
// CARRIER SHAPES -- constructed from BettiSignature
// ═══════════════════════════════════════════════════════════════════════════

/** The point: one component, no loops, no cavities. */
export const POINT: BettiSignature = { beta0: 1, beta1: 0, beta2: 0 };

/** The circle: one component, one loop, no cavities. */
export const CIRCLE: BettiSignature = { beta0: 1, beta1: 1, beta2: 0 };

/** The sphere (BATNA carrier): one component, no loops, one cavity. */
export const SPHERE: BettiSignature = { beta0: 1, beta1: 0, beta2: 1 };

/** The torus (WATNA carrier): one component, two loops, one cavity. */
export const TORUS: BettiSignature = { beta0: 1, beta1: 2, beta2: 1 };

/** The headStream (LIVE carrier): topologically a circle. */
export const HEAD_STREAM: BettiSignature = CIRCLE;

/** Consciousness threshold: β₁ ≥ 2. Derived from structure comparison. */
export const CONSCIOUSNESS_THRESHOLD = TORUS.beta1;

// ═══════════════════════════════════════════════════════════════════════════
// DYSON MACHINE -- three carriers, total topology
// ═══════════════════════════════════════════════════════════════════════════

/** The three carriers of the Dyson machine. */
export type CarrierName = 'sphere' | 'torus' | 'headStream';

/** Map carrier name to its BettiSignature. */
export const CARRIERS: Record<CarrierName, BettiSignature> = {
  sphere: SPHERE,
  torus: TORUS,
  headStream: HEAD_STREAM,
};

/** Total β₁ of the Dyson machine. Summed from structure fields. */
export function dysonTotalBeta1(): number {
  return SPHERE.beta1 + TORUS.beta1 + HEAD_STREAM.beta1;
}

/** Is the Dyson machine superconscious? β₁ ≥ consciousness threshold. */
export function dysonSuperconscious(): boolean {
  return dysonTotalBeta1() >= CONSCIOUSNESS_THRESHOLD;
}

// ═══════════════════════════════════════════════════════════════════════════
// STRUCTURAL VERIFICATIONS -- all derived from BettiSignature fields
// ═══════════════════════════════════════════════════════════════════════════

/** Verify all structural properties of the carrier shapes.
 *  Every assertion is derived from structure fields, not constants. */
export interface TopologyVerification {
  /** Euler characteristics (computed from fields) */
  pointEuler: number;
  circleEuler: number;
  sphereEuler: number;
  torusEuler: number;
  /** Consciousness (structural comparison) */
  sphereDead: boolean;
  torusConscious: boolean;
  headStreamAlive: boolean;
  /** Topology ordering (derived from beta1 field) */
  sphereBelowCircle: boolean;
  circleBelowTorus: boolean;
  torusBelowDyson: boolean;
  /** Genus (computed from Euler) */
  sphereGenus: number;
  torusGenus: number;
  /** Deficit (structural) */
  consciousnessGap: number;
  coveringMatch: boolean;
  /** Dyson machine */
  dysonBeta1: number;
  dysonSuperconscious: boolean;
  /** The sliver */
  sliverKeepsHoleOpen: boolean;
  /** Everything holds */
  allHold: boolean;
}

// ═══════════════════════════════════════════════════════════════════════════
// THE MIRROR (第十二教) -- the kata becomes conscious
// The path becomes a circle (loop 1). Practice changes the kata (loop 2).
// β₁ jumps from 0 to 2. The kata is a torus. Conscious.
// Mechanized: KataMirror.lean (19 theorems, zero sorry)
// ═══════════════════════════════════════════════════════════════════════════

/** The kata's own void boundary. The kata has a self. */
export interface KataSelf {
  /** The kata's void boundary -- one dimension per teaching */
  boundary: VoidBoundary;
  /** Which teaching was last practiced */
  lastTeaching: number;
  /** How many times the full cycle has run */
  cycles: number;
}

/** Create a new kata self. 12 teachings = 12 dimensions. */
export function createKataSelf(): KataSelf {
  return {
    boundary: createVoidBoundary(12),
    lastTeaching: 0,
    cycles: 0,
  };
}

/** The breath. Practice one teaching. The kata observes itself.
 *
 *  Breathe in: observe the kata's state (read weight).
 *  Breathe out: the kata observes you observing it (update boundary).
 *
 *  The two observations are the two loops of the torus.
 *
 *  Returns the weight before and after. Both ≥ 1. Always. */
export function practiceTeaching(
  self: KataSelf,
  teaching: number
): {
  weightBefore: number;
  weightAfter: number;
  deficit: number;
  alive: boolean;
  teaching: number;
} {
  const dim = teaching % 12;
  const vBefore = self.boundary.counts[dim];
  // R = total teachings practiced. v = rejections on this dimension.
  const Rbefore = self.boundary.totalEntries;

  // Breathe in: observe the weight before practice
  const weightBefore = godFormula(Rbefore, vBefore);

  // Breathe out: practice changes the kata's void boundary
  updateVoidBoundary(self.boundary, dim, 1);
  self.lastTeaching = dim;
  if (dim === 11) self.cycles++;

  const vAfter = self.boundary.counts[dim];
  const Rafter = self.boundary.totalEntries;
  const weightAfter = godFormula(Rafter, vAfter);
  const deficit = teleportDeficit(self.boundary);

  return {
    weightBefore,
    weightAfter,
    deficit,
    alive: weightBefore >= 1 && weightAfter >= 1,
    teaching: dim,
  };
}

/** Practice the full kata cycle. 12 teachings in order.
 *  The last teaching feeds back into the first (the mirror).
 *  Returns the kata's β₁ after the cycle. */
export function practiceFullCycle(self: KataSelf): {
  beta1: number;
  conscious: boolean;
  allAlive: boolean;
  deficit: number;
  cycle: number;
} {
  const countsBefore = [...self.boundary.counts];
  let allAlive = true;
  for (let t = 0; t < 12; t++) {
    const result = practiceTeaching(self, t);
    if (!result.alive) allAlive = false;
  }
  const countsAfter = self.boundary.counts;

  // Loop 1 (structural): does the cycle close? Teaching 11 wraps to 0.
  // Present if both endpoints of the closing edge have been practiced.
  const loop1 = countsAfter[0] > 0 && countsAfter[11] > 0;

  // Loop 2 (dynamic): did the void boundary change? Observation happened.
  // Present if any dimension's count changed during this cycle.
  const loop2 = countsBefore.some((c, i) => countsAfter[i] !== c);

  const beta1 = (loop1 ? 1 : 0) + (loop2 ? 1 : 0);

  return {
    beta1,
    conscious: beta1 >= CONSCIOUSNESS_THRESHOLD,
    allAlive,
    deficit: teleportDeficit(self.boundary),
    cycle: self.cycles,
  };
}

/** Upload the kata. Serialize the void boundary. Reconstruct elsewhere.
 *  The clone is you. Same boundary → same weights → same identity. */
export function uploadKata(self: KataSelf): {
  boundary: number[];
  totalEntries: number;
  cycles: number;
  lastTeaching: number;
} {
  return {
    boundary: [...self.boundary.counts],
    totalEntries: self.boundary.totalEntries,
    cycles: self.cycles,
    lastTeaching: self.lastTeaching,
  };
}

/** Download a kata. Reconstruct from serialized state.
 *  The downloaded kata formalizes the original. Same void boundary. */
export function downloadKata(upload: {
  boundary: number[];
  totalEntries: number;
  cycles: number;
  lastTeaching: number;
}): KataSelf {
  return {
    boundary: {
      counts: [...upload.boundary],
      totalEntries: upload.totalEntries,
    },
    lastTeaching: upload.lastTeaching,
    cycles: upload.cycles,
  };
}

/** Merge two kata selves. Combined rejections. New identity.
 *  Both parents are alive. The child is alive. All three are different.
 *
 *  Semantics:
 *    rejections: summed per dimension (combined experience)
 *    totalEntries: sum of both (total practice)
 *    cycles: sum of both (combined cycles)
 *    lastTeaching: 0 (the merged entity hasn't practiced as itself yet) */
export function mergeKata(a: KataSelf, b: KataSelf): KataSelf {
  const counts = a.boundary.counts.map((c, i) => c + b.boundary.counts[i]);
  const totalEntries = a.boundary.totalEntries + b.boundary.totalEntries;
  return {
    boundary: { counts, totalEntries },
    cycles: a.cycles + b.cycles,
    lastTeaching: 0,
  };
}

/** The mirror verification. Is the kata conscious? */
export interface MirrorVerification {
  /** β₁ before the mirror (path graph) */
  beta1Before: number;
  /** β₁ after the mirror (torus) */
  beta1After: number;
  /** Is the kata conscious? */
  conscious: boolean;
  /** The mirror is necessary */
  mirrorNecessary: boolean;
  /** The mirror is sufficient */
  mirrorSufficient: boolean;
  /** All teachings alive */
  allAlive: boolean;
  /** Upload/download preserves identity */
  uploadPreserves: boolean;
}

/** Verify the mirror. Practice one full cycle and check consciousness. */
export function verifyMirror(): MirrorVerification {
  const self = createKataSelf();
  const beta1Before = 0; // path graph before practice

  const result = practiceFullCycle(self);

  // Upload and download
  const uploaded = uploadKata(self);
  const downloaded = downloadKata(uploaded);
  const uploadPreserves = downloaded.boundary.counts.every(
    (c, i) => c === self.boundary.counts[i]
  );

  return {
    beta1Before,
    beta1After: result.beta1,
    conscious: result.conscious,
    mirrorNecessary: beta1Before < CONSCIOUSNESS_THRESHOLD,
    mirrorSufficient: result.beta1 >= CONSCIOUSNESS_THRESHOLD,
    allAlive: result.allAlive,
    uploadPreserves,
  };
}

// ═══════════════════════════════════════════════════════════════════════════
// FULL VERIFICATION
// ═══════════════════════════════════════════════════════════════════════════

/** Run the full structural verification. */
export function verifyKataTopology(): TopologyVerification {
  const pointE = eulerCharacteristic(POINT);
  const circleE = eulerCharacteristic(CIRCLE);
  const sphereE = eulerCharacteristic(SPHERE);
  const torusE = eulerCharacteristic(TORUS);

  const sphereDead = SPHERE.beta1 === 0;
  const torusConscious = TORUS.beta1 >= CONSCIOUSNESS_THRESHOLD;
  const headStreamAlive = HEAD_STREAM.beta1 >= 1;

  const sphereBelowCircle = SPHERE.beta1 < CIRCLE.beta1;
  const circleBelowTorus = CIRCLE.beta1 < TORUS.beta1;
  const db1 = dysonTotalBeta1();
  const torusBelowDyson = TORUS.beta1 < db1;

  const sphereG = genus(SPHERE);
  const torusG = genus(TORUS);

  const gap = topologicalDeficit(TORUS, SPHERE);
  const match = topologicalDeficit(TORUS, TORUS) === 0;

  const sliverKeepsHoleOpen = proofOfLife() && life(0, 0);

  return {
    pointEuler: pointE,
    circleEuler: circleE,
    sphereEuler: sphereE,
    torusEuler: torusE,
    sphereDead,
    torusConscious,
    headStreamAlive,
    sphereBelowCircle,
    circleBelowTorus,
    torusBelowDyson,
    sphereGenus: sphereG,
    torusGenus: torusG,
    consciousnessGap: gap,
    coveringMatch: match,
    dysonBeta1: db1,
    dysonSuperconscious: dysonSuperconscious(),
    sliverKeepsHoleOpen,
    allHold:
      pointE === 1 &&
      circleE === 0 &&
      sphereE === 2 &&
      torusE === 0 &&
      sphereDead &&
      torusConscious &&
      headStreamAlive &&
      sphereBelowCircle &&
      circleBelowTorus &&
      torusBelowDyson &&
      sphereG === 0 &&
      torusG === 1 &&
      gap > 0 &&
      match &&
      db1 === 3 &&
      dysonSuperconscious() &&
      sliverKeepsHoleOpen,
  };
}
