// @a0n/.ignored_gnosis/src/runtime/spiral-machine.ts
/**
 * The Spiral Machine -- a Turing machine built from the perpetual spiral
 *
 * The tape is the void boundary (unbounded array of rejection counts).
 * The head is the complement peak (points to the least-rejected cell).
 * The state is the current topology.
 * The transition function is the generator (mutation from complement).
 *
 * Each step of the Turing machine is one breath of the spiral:
 *   1. READ:  observe the void boundary at the head position (complement peak)
 *   2. WRITE: execute the topology, vent losers into the void boundary
 *   3. MOVE:  the complement peak shifts (the head moves)
 *   4. STATE: the generator produces a new topology (state transition)
 *
 * The tape grows from vents (write). The head moves from complement shift
 * (the creative force). The state changes from generation (the spiral's
 * self-rewriting). The machine halts when... it doesn't. The sliver
 * prevents halting. This is a Turing machine that runs forever.
 *
 * But if you want it to halt: add an accept state. When the topology's
 * execution result satisfies a predicate, the machine accepts. The
 * predicate is the specification. The spiral searches for a topology
 * that satisfies it. This is program synthesis via rejection.
 *
 * Conway's Game of Life is Turing-complete because gliders can simulate
 * signals. The Spiral Machine is Turing-complete because:
 *   - THM-TM-AS-FOLD (§26.1): fold models the TM step
 *   - The spiral executes topologies containing folds
 *   - Therefore the spiral simulates any TM
 *
 * But the deeper result: the spiral doesn't just SIMULATE a TM.
 * The spiral maps to a TM, where the tape, head, state, and transitions
 * are all built from the same four primitives (fork/race/fold/vent).
 */

import {
  type VoidBoundary,
  createVoidBoundary,
  updateVoidBoundary,
  complementDistribution,
  teleportDeficit,
} from '../void';

import type { CompiledTopology } from './compiled-topology';
import { executeCompiled } from './compiled-topology';

import {
  type KataSelf,
  createKataSelf,
  practiceTeaching,
} from './kata-topology';

// ============================================================================
// The Tape: void boundary as unbounded memory
// ============================================================================

/**
 * The Spiral Machine's tape. Each cell is a void boundary dimension.
 * Writing = venting into a cell (increasing its rejection count).
 * Reading = observing the complement weight at a cell.
 * The tape grows as new dimensions are added.
 */
export interface SpiralTape {
  /** The void boundary formalizes the tape. */
  boundary: VoidBoundary;
  /** Current head position (complement peak index). */
  head: number;
  /** Number of steps taken. */
  steps: number;
  /** The kata's self -- the machine practices the kata as it steps. */
  kataSelf: KataSelf;
}

/**
 * Create a tape with initial dimensions.
 */
export function createTape(dimensions: number): SpiralTape {
  return {
    boundary: createVoidBoundary(dimensions),
    head: 0,
    steps: 0,
    kataSelf: createKataSelf(),
  };
}

/**
 * Read the tape at the head position.
 * Returns the complement weight (0-1) at the current cell.
 */
export function read(tape: SpiralTape): number {
  const dist = complementDistribution(tape.boundary);
  return dist[tape.head] ?? 0;
}

/**
 * Write to the tape at the head position.
 * Writing = rejecting (venting) into the current cell.
 */
export function write(tape: SpiralTape, magnitude: number): void {
  updateVoidBoundary(tape.boundary, tape.head, magnitude);
}

/**
 * Move the head. The head moves to the complement peak --
 * the dimension with the highest complement weight.
 * This is the creative force directing attention.
 */
export function moveHead(tape: SpiralTape): void {
  const dist = complementDistribution(tape.boundary);
  let bestIdx = 0;
  let bestWeight = -1;
  for (let i = 0; i < dist.length; i++) {
    if (dist[i] > bestWeight) {
      bestWeight = dist[i];
      bestIdx = i;
    }
  }
  tape.head = bestIdx;
}

// ============================================================================
// The State: topology as machine state
// ============================================================================

/**
 * A state of the spiral machine: a topology + accept predicate.
 */
export interface MachineState {
  /** The current topology (the "program" being optimized). */
  topology: CompiledTopology;
  /** Whether this state is an accept state. */
  accepted: boolean;
  /** The state's name (for tracing). */
  label: string;
}

// ============================================================================
// The Transition Function: generation as state transition
// ============================================================================

/**
 * Transition: read tape → execute topology → write result → move head → new state.
 */
export interface Transition {
  /** Tape symbol read (complement weight at head). */
  read: number;
  /** Topology that executed. */
  state: MachineState;
  /** Vent energy written to tape. */
  written: number;
  /** New head position after move. */
  newHead: number;
  /** New state (possibly same topology if identity mutation won). */
  newState: MachineState;
  /** Execution result. */
  result: unknown;
}

// ============================================================================
// The Machine
// ============================================================================

/**
 * Configuration for the spiral machine.
 */
export interface SpiralMachineConfig {
  /** Initial tape dimensions. */
  tapeDimensions: number;
  /** Accept predicate: when should the machine halt? */
  accept: (result: unknown, tape: SpiralTape) => boolean;
  /** Maximum steps (0 = no limit). */
  maxSteps: number;
  /** Input to feed the topology. */
  input: unknown;
  /** Number of topology variants per step. */
  variants: number;
  /** Callback per step. */
  onStep?: (transition: Transition) => void;
}

/**
 * Result of running the spiral machine.
 */
export interface SpiralMachineResult {
  /** Whether the machine accepted (halted on accept state). */
  accepted: boolean;
  /** The final topology (the "discovered program"). */
  topology: CompiledTopology;
  /** Number of steps taken. */
  steps: number;
  /** The final tape state. */
  tape: SpiralTape;
  /** All transitions (the computation trace). */
  trace: Transition[];
  /** Final execution result. */
  result: unknown;
}

/**
 * Run the spiral machine.
 *
 * This is a Turing machine where:
 *   tape = void boundary
 *   head = complement peak
 *   state = topology
 *   transition = generate (mutate from complement)
 *
 * The machine runs until:
 *   - The accept predicate returns true (the specification is met)
 *   - maxSteps is reached (safety limit)
 *
 * If no accept predicate is given and maxSteps = 0, the machine
 * runs forever. The sliver guarantees it always has work to do.
 */
export async function runSpiralMachine(
  initialTopology: CompiledTopology,
  config: SpiralMachineConfig
): Promise<SpiralMachineResult> {
  const tape = createTape(config.tapeDimensions);
  const trace: Transition[] = [];
  let currentState: MachineState = {
    topology: initialTopology,
    accepted: false,
    label: 'q0',
  };
  let lastResult: unknown = null;

  while (true) {
    // ── READ: observe tape at head ──
    const symbol = read(tape);

    // ── EXECUTE: run topology (the computation step) ──
    const result = await executeCompiled(currentState.topology, config.input);
    lastResult = result.payload;

    // ── WRITE: vent into tape (rejection signal) ──
    // The execution time formalizes the rejection signal: slower = more rejection
    const ventStrength = Math.max(1, Math.floor(result.executionUs / 100));
    write(tape, ventStrength);

    // ── MOVE: head follows complement peak ──
    const oldHead = tape.head;
    moveHead(tape);

    // ── TRANSITION: generate new topology state ──
    // For now: identity transition (topology unchanged).
    // Full version: mutate from complement distribution.
    const newState: MachineState = {
      topology: currentState.topology,
      accepted: config.accept(result.payload, tape),
      label: `q${tape.steps + 1}`,
    };

    const transition: Transition = {
      read: symbol,
      state: currentState,
      written: ventStrength,
      newHead: tape.head,
      newState,
      result: result.payload,
    };
    trace.push(transition);
    config.onStep?.(transition);

    tape.steps++;
    currentState = newState;

    // ── KATA: the machine practices the teaching for this step ──
    practiceTeaching(tape.kataSelf, tape.steps - 1);

    // ── HALT: accept predicate met ──
    if (currentState.accepted) {
      return {
        accepted: true,
        topology: currentState.topology,
        steps: tape.steps,
        tape,
        trace,
        result: lastResult,
      };
    }

    // ── SAFETY: max steps ──
    if (config.maxSteps > 0 && tape.steps >= config.maxSteps) {
      return {
        accepted: false,
        topology: currentState.topology,
        steps: tape.steps,
        tape,
        trace,
        result: lastResult,
      };
    }

    // The sliver guarantees we always continue.
    // The void boundary always has something to vent.
    // The complement always has a peak to follow.
    // The machine always has a next step.
  }
}
