// @a0n/.ignored_gnosis/src/runtime/worker-runtime.ts
import type { GraphAST } from '../betty/compiler';
import {
  compileSyncForkTopology,
  compileTopology,
  executeCompiled,
  executeCompiledSyncFork,
  type CompiledExecutionResult,
  type CompiledSyncForkTopology,
  type CompiledTopology,
  type SyncHandler,
} from './compiled-topology';
import {
  GnosisNativeRuntime,
  type GnosisNativeRuntimeSnapshot,
} from './native-runtime';
import type { GnosisRegistry } from './registry';

export interface GnosisWorkerRuntimeOptions {
  readonly ast: GraphAST;
  readonly registry: GnosisRegistry;
  readonly fingerprint: string;
  readonly nativeControlPlane?: boolean;
  readonly syncHandlerOverrides?: Map<string, SyncHandler>;
}

export interface GnosisWorkerRuntimeLoadResult {
  readonly runtime: GnosisWorkerRuntime;
  readonly nativeSnapshot: GnosisNativeRuntimeSnapshot | null;
}

export interface GnosisWorkerExecutionResult extends CompiledExecutionResult {
  readonly nativeControlPlane: boolean;
}

/**
 * Worker-oriented GG runtime.
 *
 * The load path admits and lowers the control plane once. The hot path executes
 * a precompiled handler plan without AST traversal, label lookup, or log
 * construction. When nativeControlPlane is enabled, FORK/RACE/FOLD/VENT edges
 * are also fed through the Rust/WASM frame runtime at load time so the worker
 * carries the same topology metrics as the native path without paying that
 * cost per request.
 */
export class GnosisWorkerRuntime {
  private constructor(
    private readonly compiled: CompiledTopology,
    private readonly compiledSyncFork: CompiledSyncForkTopology | null,
    private readonly nativeSnapshot: GnosisNativeRuntimeSnapshot | null
  ) {}

  static async load(
    options: GnosisWorkerRuntimeOptions
  ): Promise<GnosisWorkerRuntimeLoadResult> {
    const compiled = compileTopology(
      options.ast,
      options.registry,
      options.fingerprint
    );
    const compiledSyncFork = options.syncHandlerOverrides
      ? compileSyncForkTopology(
          options.ast,
          options.registry,
          options.fingerprint,
          options.syncHandlerOverrides
        )
      : null;
    let nativeSnapshot: GnosisNativeRuntimeSnapshot | null = null;

    if (options.nativeControlPlane !== false) {
      const nativeRuntime = new GnosisNativeRuntime();
      nativeSnapshot = await nativeRuntime.processEdges(options.ast.edges);
    }

    const runtime = new GnosisWorkerRuntime(
      compiled,
      compiledSyncFork,
      nativeSnapshot
    );
    return { runtime, nativeSnapshot };
  }

  async execute(input: unknown): Promise<GnosisWorkerExecutionResult> {
    const result = this.compiledSyncFork
      ? executeCompiledSyncFork(this.compiledSyncFork, input)
      : await executeCompiled(this.compiled, input);
    return {
      ...result,
      nativeControlPlane: this.nativeSnapshot !== null,
    };
  }

  snapshot(): GnosisNativeRuntimeSnapshot | null {
    return this.nativeSnapshot;
  }

  get nodeCount(): number {
    return this.compiled.nodeCount;
  }

  get hasForks(): boolean {
    return this.compiled.hasForks;
  }
}

/**
 * Loads the Gnosis Worker Runtime.
 */
export async function loadGnosisWorkerRuntime(
  options: GnosisWorkerRuntimeOptions
): Promise<GnosisWorkerRuntimeLoadResult> {
  return GnosisWorkerRuntime.load(options);
}
