// @a0n/.ignored_gnosis/forkjoin-app/client/components/3D/SafeEffects.tsx
/**
 * SafeEffects - Wrapper for @react-three/postprocessing that handles WebGL context loss
 *
 * The EffectComposer can crash when WebGL context is lost (common on mobile/low-power devices).
 * This wrapper catches those errors and degrades gracefully.
 */

import { Component, ReactNode, useState, useEffect } from '@a0n/raect';
import { useThree } from '@a0n/aeon-3d/fiber';
import { EffectComposer, Bloom } from '@a0n/aeon-3d/postprocessing';

interface ErrorBoundaryProps {
  children: ReactNode;
  fallback?: ReactNode;
}

interface ErrorBoundaryState {
  hasError: boolean;
}

/**
 * Error boundary specifically for Three.js effects
 */
export class EffectErrorBoundary extends Component<
  ErrorBoundaryProps,
  ErrorBoundaryState
> {
  constructor(props: ErrorBoundaryProps) {
    super(props);
    this.state = { hasError: false };
  }

  static getDerivedStateFromError(): ErrorBoundaryState {
    return { hasError: true };
  }

  componentDidCatch(error: Error): void {
    // Silently handle WebGL context loss errors
    if (
      error.message?.includes('alpha') ||
      error.message?.includes('Context')
    ) {
      console.debug('[Effects] WebGL context unavailable, effects disabled');
    }
  }

  render(): ReactNode {
    if (this.state.hasError) {
      return this.props.fallback ?? null;
    }
    return this.props.children;
  }
}

interface SafeBloomProps {
  intensity?: number;
  luminanceThreshold?: number;
  luminanceSmoothing?: number;
}

/**
 * Safe Bloom effect that checks for WebGL context availability
 */
function SafeBloomInner({
  intensity = 0.5,
  luminanceThreshold = 0.3,
  luminanceSmoothing = 0.9,
}: SafeBloomProps) {
  const { gl } = useThree();
  const [contextValid, setContextValid] = useState(false);
  const [mounted, setMounted] = useState(false);

  useEffect(() => {
    // Delay mounting to avoid race conditions with WebGL initialization
    // Use a longer delay to ensure the context is fully ready
    const timer = setTimeout(() => {
      try {
        // Check if renderer and its properties are valid
        if (!gl || !gl.domElement) {
          setContextValid(false);
          setMounted(true);
          return;
        }

        // Try to get the WebGL context directly from the canvas
        const canvas = gl.domElement;
        const webglContext =
          canvas.getContext('webgl2') || canvas.getContext('webgl');

        // Check if we have a valid, non-lost context
        const isValid = webglContext && !webglContext.isContextLost();

        // Also verify renderer state
        const rendererValid = gl.info && gl.capabilities;

        setContextValid(!!isValid && !!rendererValid);
        setMounted(true);
      } catch {
        console.debug('[Effects] WebGL context check failed, effects disabled');
        setContextValid(false);
        setMounted(true);
      }
    }, 300); // Increased delay for more stability

    return () => clearTimeout(timer);
  }, [gl]);

  useEffect(() => {
    if (!mounted || !gl?.domElement) return;

    const canvas = gl.domElement;
    const handleContextLost = () => {
      console.debug('[Effects] WebGL context lost, disabling effects');
      setContextValid(false);
    };
    const handleContextRestored = () => {
      console.debug('[Effects] WebGL context restored');
      setContextValid(true);
    };

    canvas.addEventListener('webglcontextlost', handleContextLost);
    canvas.addEventListener('webglcontextrestored', handleContextRestored);

    return () => {
      canvas.removeEventListener('webglcontextlost', handleContextLost);
      canvas.removeEventListener('webglcontextrestored', handleContextRestored);
    };
  }, [gl, mounted]);

  // Don't render until we've checked the context and it's valid
  if (!mounted || !contextValid) {
    return null;
  }

  // Double-check context is still valid before rendering effects
  try {
    const canvas = gl?.domElement;
    if (!canvas) return null;

    const ctx = canvas.getContext('webgl2') || canvas.getContext('webgl');
    if (!ctx || ctx.isContextLost()) return null;
  } catch {
    return null;
  }

  return (
    <EffectComposer multisampling={0}>
      <Bloom
        intensity={intensity}
        luminanceThreshold={luminanceThreshold}
        luminanceSmoothing={luminanceSmoothing}
      />
    </EffectComposer>
  );
}

/**
 * SafeBloom - Bloom effect with automatic context loss handling
 */
export function SafeBloom(props: SafeBloomProps) {
  return (
    <EffectErrorBoundary>
      <SafeBloomInner {...props} />
    </EffectErrorBoundary>
  );
}
