// @a0n/.ignored_gnosis/forkjoin-app/client/components/Stack3D/StackVisualization.tsx
/**
 * StackVisualization - 3D Pyramid of the Aeon Foundation stack
 *
 * Uses simple box geometries without drei Text to avoid font-related crashes.
 * Labels are rendered as HTML overlay with positions synced to 3D geometry.
 */

import { useRef, useState, useEffect, useCallback } from '@a0n/raect';
import { Canvas, useFrame, useThree } from '@a0n/aeon-3d/fiber';
import { Float, Edges } from '@a0n/aeon-3d/drei';
import * as THREE from '@a0n/aeon-3d/three';
import {
  STACK_LAYERS_TOP_DOWN,
  type StackLayerDefinition,
  type StackSegmentDefinition,
} from '../stack-model';

interface StackLayer {
  id: string;
  name: string;
  layer: string;
  color: string;
  index: number;
  href: string;
  segments?: StackLayerSegment[];
}

interface StackLayerSegment {
  name: string;
  ratio: number;
  color: string;
}

// Render order is bottom (foundation) to top (apex).
const stackLayers: StackLayer[] = (
  STACK_LAYERS_TOP_DOWN as readonly StackLayerDefinition[]
)
  .slice()
  .reverse()
  .map((layer: StackLayerDefinition, index) => ({
    id: layer.id,
    name: layer.name,
    layer: layer.layer,
    color: layer.color,
    index,
    href: layer.href,
    segments: layer.segments?.map((segment: StackSegmentDefinition) => ({
      name: segment.label,
      ratio: segment.ratio,
      color: segment.color,
    })),
  }));

interface StackGeometry {
  baseSize: number;
  gap: number;
  groupY: number;
  layerHeight: number;
  stackSpan: number;
  taper: number;
  wrapperHeight: number;
}

function getStackGeometry(totalLayers: number): StackGeometry {
  const compact = totalLayers > 16;
  const layerHeight = compact ? 0.44 : 0.65;
  const gap = compact ? 0.05 : 0.12;
  const baseSize = compact ? 6.2 : 4.5;
  const taper = compact ? 0.62 : 0.7;
  const stackSpan = Math.max(0, totalLayers - 1) * (layerHeight + gap);

  return {
    baseSize,
    gap,
    groupY: -stackSpan / 2,
    layerHeight,
    stackSpan,
    taper,
    wrapperHeight: Math.max(580, 360 + totalLayers * 22),
  };
}

function getLayerSize(
  index: number,
  totalLayers: number,
  geometry: StackGeometry
): number {
  const progress = totalLayers <= 1 ? 0 : index / (totalLayers - 1);
  return geometry.baseSize * (1 - progress * geometry.taper);
}

interface LabelPosition {
  x: number;
  y: number;
}

/**
 * Renders the Stack Visualization view.
 */
export function StackVisualization() {
  const [isVisible, setIsVisible] = useState(false);
  const [labelPositions, setLabelPositions] = useState<LabelPosition[]>([]);
  const containerRef = useRef<HTMLDivElement>(null);
  const geometry = getStackGeometry(stackLayers.length);

  // Only mount when in viewport
  useEffect(() => {
    if (!containerRef.current) return;

    const observer = new IntersectionObserver(
      ([entry]) => {
        setIsVisible(entry.isIntersecting);
      },
      { threshold: 0.1, rootMargin: '100px' }
    );

    observer.observe(containerRef.current);
    return () => observer.disconnect();
  }, []);

  const handlePositionsUpdate = useCallback((positions: LabelPosition[]) => {
    setLabelPositions(positions);
  }, []);

  return (
    <div ref={containerRef} className="stack-viz-wrapper">
      {isVisible && (
        <Canvas
          camera={{
            position: [
              geometry.baseSize * 2.4,
              Math.max(5, geometry.stackSpan * 0.58 + 2.8),
              geometry.baseSize * 3,
            ],
            fov: stackLayers.length > 18 ? 32 : 40,
          }}
          gl={{
            antialias: true,
            alpha: true,
            powerPreference: 'high-performance',
          }}
          dpr={[1, 1.5]}
        >
          <ambientLight intensity={0.3} />
          <pointLight position={[15, 15, 15]} intensity={0.7} />
          <pointLight
            position={[-10, 5, -10]}
            intensity={0.2}
            color="#336699"
          />
          <directionalLight position={[5, 10, 5]} intensity={0.3} />

          <PyramidScene
            geometry={geometry}
            onPositionsUpdate={handlePositionsUpdate}
          />
        </Canvas>
      )}

      {/* HTML labels overlay - clickable links to sections */}
      <div className="pyramid-labels">
        {stackLayers.map((layer, idx) => {
          const pos = labelPositions[idx];
          // Fallback to percentage-based position if 3D positions not yet available
          const yPercent = pos
            ? undefined
            : 90 - idx * (80 / Math.max(1, stackLayers.length - 1));
          return (
            <a
              key={layer.id}
              href={layer.href}
              className="pyramid-label"
              style={
                {
                  top: pos ? `${pos.y}px` : `${yPercent}%`,
                  '--label-color': layer.color,
                  '--border-width': `${2 + idx * 0.8}px`,
                  opacity: pos ? 1 : 0.8,
                } as React.CSSProperties
              }
            >
              <span className="label-name">{layer.name}</span>
              <span className="label-divider">·</span>
              <span className="label-type">{layer.layer}</span>
            </a>
          );
        })}
      </div>

      <style>{`
        .stack-viz-wrapper {
          position: relative;
          width: 100%;
          height: ${geometry.wrapperHeight}px;
        }

        .pyramid-labels {
          position: absolute;
          right: 8%;
          top: 0;
          bottom: 0;
        }

        .pyramid-label {
          position: absolute;
          right: 0;
          display: flex;
          flex-direction: row;
          align-items: center;
          gap: 8px;
          transform: translateY(-50%);
          padding: 6px 14px;
          background: linear-gradient(270deg, rgba(0,0,0,0.85) 0%, rgba(0,0,0,0.6) 70%, transparent 100%);
          border-right: var(--border-width) solid var(--label-color);
          backdrop-filter: blur(4px);
          white-space: nowrap;
          transition: top 0.1s ease-out, background 0.2s ease;
          cursor: pointer;
          text-decoration: none;
        }

        .pyramid-label:hover {
          background: linear-gradient(270deg, rgba(0,0,0,0.95) 0%, rgba(0,0,0,0.8) 70%, rgba(0,0,0,0.2) 100%);
        }

        .pyramid-label:hover .label-name {
          color: var(--label-color);
        }

        .label-name {
          font-size: 1rem;
          font-weight: 700;
          color: #ffffff;
          text-shadow: 0 1px 4px rgba(0,0,0,0.8);
        }

        .label-divider {
          color: var(--label-color);
          opacity: 0.6;
        }

        .label-type {
          font-size: 0.7rem;
          text-transform: uppercase;
          letter-spacing: 0.1em;
          color: var(--label-color);
          font-weight: 500;
        }

        @media (max-width: 768px) {
          .stack-viz-wrapper {
            height: ${Math.max(520, geometry.wrapperHeight - 120)}px;
          }

          .pyramid-labels {
            right: 2%;
          }

          .pyramid-label {
            padding: 6px 12px;
          }

          .label-name {
            font-size: 0.9rem;
          }

          .label-type {
            font-size: 0.55rem;
          }
        }
      `}</style>
    </div>
  );
}

interface PyramidSceneProps {
  geometry: StackGeometry;
  onPositionsUpdate: (positions: LabelPosition[]) => void;
}

function PyramidScene({ geometry, onPositionsUpdate }: PyramidSceneProps) {
  const groupRef = useRef<THREE.Group>(null);
  const layerRefs = useRef<(THREE.Object3D | null)[]>([]);
  const frameCount = useRef(0);
  const { camera, size } = useThree();

  // Project 3D positions to 2D screen coordinates (throttled to every 3 frames)
  useFrame(() => {
    if (!groupRef.current) return;

    // Throttle updates to reduce React re-renders
    frameCount.current++;
    if (frameCount.current % 3 !== 0) return;

    const positions: LabelPosition[] = [];
    const tempVec = new THREE.Vector3();

    for (let i = 0; i < stackLayers.length; i++) {
      const layerObject = layerRefs.current[i];
      if (layerObject) {
        // Get world position of the mesh
        layerObject.getWorldPosition(tempVec);
        // Offset to the right edge of the layer
        const layerSize = getLayerSize(i, stackLayers.length, geometry);
        tempVec.x += layerSize / 2 + 0.7;

        // Project to screen coordinates
        tempVec.project(camera as unknown as THREE.Camera);

        // Convert to pixel coordinates
        const x = (tempVec.x * 0.5 + 0.5) * size.width;
        const y = (-tempVec.y * 0.5 + 0.5) * size.height;

        positions.push({ x, y });
      }
    }

    if (positions.length === stackLayers.length) {
      onPositionsUpdate(positions);
    }
  });

  const setLayerRef = (index: number) => (ref: THREE.Object3D | null) => {
    layerRefs.current[index] = ref;
  };

  return (
    <group ref={groupRef} position={[0, geometry.groupY, 0]}>
      {stackLayers.map((layer) => (
        <PyramidLayer
          key={layer.id}
          geometry={geometry}
          layer={layer}
          totalLayers={stackLayers.length}
          meshRef={setLayerRef(layer.index)}
        />
      ))}
    </group>
  );
}

interface PyramidLayerProps {
  geometry: StackGeometry;
  layer: StackLayer;
  totalLayers: number;
  meshRef: (ref: THREE.Object3D | null) => void;
}

function PyramidLayer({
  geometry,
  layer,
  totalLayers,
  meshRef,
}: PyramidLayerProps) {
  const internalRef = useRef<THREE.Group>(null);

  const yPosition = layer.index * (geometry.layerHeight + geometry.gap);
  const layerSize = getLayerSize(layer.index, totalLayers, geometry);

  // Expose ref to parent for position tracking
  useEffect(() => {
    meshRef(internalRef.current);
    return () => meshRef(null);
  }, [meshRef]);

  useFrame((state) => {
    if (!internalRef.current) return;
    const time = state.clock.elapsedTime;

    const waveOffset = layer.index * 0.4;
    const breathe = Math.sin(time * 2 - waveOffset) * 0.5 + 0.5;

    const scaleAmount = 1 + breathe * 0.015;
    internalRef.current.scale.set(scaleAmount, 1, scaleAmount);

    const hover = Math.sin(time * 1.5 - waveOffset * 0.5) * 0.03;
    internalRef.current.position.y = hover;

    const helixPhase = (layer.index / totalLayers) * Math.PI * 2;
    internalRef.current.rotation.y = Math.sin(time * 0.5 + helixPhase) * 0.15;

    const emissiveIntensity = 0.2 + breathe * 0.3;
    internalRef.current.traverse((child) => {
      if (child instanceof THREE.Mesh) {
        const material = child.material;
        if (material instanceof THREE.MeshPhysicalMaterial) {
          material.emissiveIntensity = emissiveIntensity;
        }
      }
    });
  });

  const segments: StackLayerSegment[] = layer.segments ?? [
    { name: layer.name, ratio: 1, color: layer.color },
  ];
  const totalRatio = segments.reduce((sum, segment) => sum + segment.ratio, 0);
  let xCursor = -layerSize / 2;

  return (
    <Float speed={0.8} rotationIntensity={0} floatIntensity={0.02}>
      <group position={[0, yPosition, 0]}>
        <group ref={internalRef}>
          {segments.map((segment) => {
            const normalizedRatio =
              totalRatio > 0 ? segment.ratio / totalRatio : 1 / segments.length;
            const segmentWidth = layerSize * normalizedRatio;
            const xCenter = xCursor + segmentWidth / 2;
            xCursor += segmentWidth;

            return (
              <mesh
                key={`${layer.name}-${segment.name}`}
                position={[xCenter, 0, 0]}
              >
                <boxGeometry
                  args={[segmentWidth, geometry.layerHeight, layerSize]}
                />
                <meshPhysicalMaterial
                  color={segment.color}
                  emissive={segment.color}
                  emissiveIntensity={0.25}
                  metalness={0.9}
                  roughness={0.1}
                  transparent
                  opacity={0.8}
                  envMapIntensity={0.5}
                />
                <Edges threshold={15} color={segment.color} />
              </mesh>
            );
          })}
        </group>
      </group>
    </Float>
  );
}
