// @a0n/gnosis/src/betti/ast-equivalence.ts
/**
 * ast-equivalence.ts -- Canonical AST comparison for bootstrap verification.
 *
 * Two GraphASTs are structurally equivalent if, after canonical serialization
 * (sorted nodes by ID, sorted edges by source+target+type, normalized
 * property key order), they produce identical structures.
 */

import type { GraphAST, ASTNode, ASTEdge } from '../betty/compiler';

// ============================================================================
// Canonical serialization
// ============================================================================

interface CanonicalNode {
  id: string;
  labels: string[];
  properties: [string, string][];
}

interface CanonicalEdge {
  sourceIds: string[];
  targetIds: string[];
  type: string;
  properties: [string, string][];
}

interface CanonicalAST {
  nodes: CanonicalNode[];
  edges: CanonicalEdge[];
}

function canonicalizeNode(node: ASTNode): CanonicalNode {
  return {
    id: node.id,
    labels: [...node.labels].sort(),
    properties: Object.entries(node.properties).sort(([a], [b]) =>
      a.localeCompare(b)
    ),
  };
}

function canonicalizeEdge(edge: ASTEdge): CanonicalEdge {
  return {
    sourceIds: [...edge.sourceIds].sort(),
    targetIds: [...edge.targetIds].sort(),
    type: edge.type,
    properties: Object.entries(edge.properties).sort(([a], [b]) =>
      a.localeCompare(b)
    ),
  };
}

function edgeSortKey(edge: CanonicalEdge): string {
  return `${edge.sourceIds.join(',')}|${edge.targetIds.join(',')}|${edge.type}`;
}

/**
 * Produce a canonical representation of a GraphAST.
 * Nodes sorted by ID, edges sorted by source+target+type,
 * properties sorted by key.
 */
export function canonicalize(ast: GraphAST): CanonicalAST {
  const nodes = Array.from(ast.nodes.values())
    .map(canonicalizeNode)
    .sort((a, b) => a.id.localeCompare(b.id));

  const edges = ast.edges
    .map(canonicalizeEdge)
    .sort((a, b) => edgeSortKey(a).localeCompare(edgeSortKey(b)));

  return { nodes, edges };
}

/**
 * Serialize a canonical AST to a stable JSON string.
 */
export function serializeCanonical(ast: GraphAST): string {
  return JSON.stringify(canonicalize(ast));
}

// ============================================================================
// Structural equivalence
// ============================================================================

/**
 * Check if two GraphASTs are structurally equivalent.
 */
export function areStructurallyEquivalent(a: GraphAST, b: GraphAST): boolean {
  return serializeCanonical(a) === serializeCanonical(b);
}

// ============================================================================
// Diff report
// ============================================================================

export interface ASTDiffEntry {
  kind:
    | 'node-missing'
    | 'node-extra'
    | 'node-differs'
    | 'edge-missing'
    | 'edge-extra'
    | 'edge-differs';
  id: string;
  details?: string;
}

/**
 * Produce a diff report between two GraphASTs.
 * Returns an empty array if they are equivalent.
 */
export function diffASTs(expected: GraphAST, actual: GraphAST): ASTDiffEntry[] {
  const diffs: ASTDiffEntry[] = [];

  const canonExpected = canonicalize(expected);
  const canonActual = canonicalize(actual);

  // Node diffs
  const expectedNodeMap = new Map(canonExpected.nodes.map((n) => [n.id, n]));
  const actualNodeMap = new Map(canonActual.nodes.map((n) => [n.id, n]));

  for (const [id, node] of expectedNodeMap) {
    const actualNode = actualNodeMap.get(id);
    if (!actualNode) {
      diffs.push({ kind: 'node-missing', id });
    } else if (JSON.stringify(node) !== JSON.stringify(actualNode)) {
      diffs.push({
        kind: 'node-differs',
        id,
        details: `expected ${JSON.stringify(node)}, got ${JSON.stringify(
          actualNode
        )}`,
      });
    }
  }

  for (const [id] of actualNodeMap) {
    if (!expectedNodeMap.has(id)) {
      diffs.push({ kind: 'node-extra', id });
    }
  }

  // Edge diffs
  const expectedEdgeKeys = new Set(canonExpected.edges.map(edgeSortKey));
  const actualEdgeKeys = new Set(canonActual.edges.map(edgeSortKey));

  for (const key of expectedEdgeKeys) {
    if (!actualEdgeKeys.has(key)) {
      diffs.push({ kind: 'edge-missing', id: key });
    }
  }

  for (const key of actualEdgeKeys) {
    if (!expectedEdgeKeys.has(key)) {
      diffs.push({ kind: 'edge-extra', id: key });
    }
  }

  // Check edge property differences for matching keys
  const expectedEdgeMap = new Map(
    canonExpected.edges.map((e) => [edgeSortKey(e), e])
  );
  const actualEdgeMap = new Map(
    canonActual.edges.map((e) => [edgeSortKey(e), e])
  );

  for (const [key, expectedEdge] of expectedEdgeMap) {
    const actualEdge = actualEdgeMap.get(key);
    if (
      actualEdge &&
      JSON.stringify(expectedEdge) !== JSON.stringify(actualEdge)
    ) {
      diffs.push({
        kind: 'edge-differs',
        id: key,
        details: `properties differ`,
      });
    }
  }

  return diffs;
}
