// @a0n/gnosis/src/betti/boundary-topology.ts
/**
 * boundary-topology.ts -- TypeScript mirror of TopologicalLucasDynamics.lean.
 *
 * Mirrors:
 *   - BoundaryTopology (Open | Closed)
 *   - IsomorphicManifold
 *   - topologicalKernel (fib / lucas dispatch)
 *   - topologicalGap (L_n - F_n = 2·F_{n-1})
 *   - √5 convergence ratio (L_n/F_n → φ + 1/φ)
 *   - Pell discriminant (L_n² = 5·F_n² ± 4)
 *   - Product doubling (F_n·L_n = F_{2n})
 *   - Reynolds-like friction bound (φ² ≈ 2.618)
 *
 * The Lean proofs verify these identities via native_decide;
 * this module is the runtime that the Betti compiler calls into.
 */

// ============================================================================
// Constants
// ============================================================================

/** The Golden Ratio: φ = (1 + √5) / 2 ≈ 1.618... */
export const PHI = (1 + Math.sqrt(5)) / 2;

/** The Golden Vent: 1/φ = φ - 1 ≈ 0.618... */
export const PHI_INV = PHI - 1;

/** √5 = φ + 1/φ ≈ 2.236... — the convergence limit of L_n/F_n. */
export const SQRT5 = Math.sqrt(5);

/**
 * The critical friction limit: φ².
 * When the Reynolds number of the Betti manifold exceeds this,
 * the laminar execution layers shear and β₁ collapses.
 */
export const PHI_SQUARED = PHI * PHI;

// ============================================================================
// Sequences (match Lean definitions exactly)
// ============================================================================

/**
 * Fibonacci sequence: F(0) = 0, F(1) = 1, F(n+2) = F(n+1) + F(n).
 * Matches TopologicalLucasDynamics.fib.
 */
export function fib(n: number): number {
  if (n < 0) throw new Error(`[BoundaryTopology] fib(${n}): negative index`);
  let a = 0;
  let b = 1;
  for (let i = 0; i < n; i++) {
    const next = a + b;
    a = b;
    b = next;
  }
  return a;
}

/**
 * Lucas sequence: L(0) = 2, L(1) = 1, L(n+2) = L(n+1) + L(n).
 * Matches TopologicalLucasDynamics.lucas.
 */
export function lucas(n: number): number {
  if (n < 0) throw new Error(`[BoundaryTopology] lucas(${n}): negative index`);
  let a = 2;
  let b = 1;
  for (let i = 0; i < n; i++) {
    const next = a + b;
    a = b;
    b = next;
  }
  return a;
}

// ============================================================================
// BoundaryTopology -- the type-level discriminant
// ============================================================================

/**
 * Boundary condition of a topological manifold.
 * Mirrors TopologicalLucasDynamics.BoundaryTopology.
 *
 *   Open   → linear mesh, F_n states (path graph)
 *   Closed → cyclic mesh, L_n states (cycle graph)
 */
export type BoundaryTopology = 'Open' | 'Closed';

/**
 * Topological kernel: dispatches to fib or lucas based on boundary.
 * Mirrors TopologicalLucasDynamics.topologicalKernel.
 */
export function topologicalKernel(
  n: number,
  boundary: BoundaryTopology
): number {
  return boundary === 'Open' ? fib(n) : lucas(n);
}

// ============================================================================
// IsomorphicManifold -- the compile-time type
// ============================================================================

/**
 * An IsomorphicManifold binds a domain name, a dimension, and a boundary
 * condition. The Betti compiler uses this to enforce boundary safety:
 * any manifold marked Closed evaluates through the Lucas kernel.
 *
 * Mirrors TopologicalLucasDynamics.IsomorphicManifold.
 */
export interface IsomorphicManifold {
  /** Human-readable domain name (e.g., "1D Ising Closed Ring"). */
  readonly domainName: string;
  /** Dimension of the manifold (mesh size / index parameter). */
  readonly dimension: number;
  /** Boundary condition: Open (linear) or Closed (cyclic). */
  readonly boundary: BoundaryTopology;
}

/**
 * Evaluate the state-space cardinality of a manifold.
 * Mirrors TopologicalLucasDynamics.evaluateStateSpace.
 */
export function evaluateStateSpace(m: IsomorphicManifold): number {
  return topologicalKernel(m.dimension, m.boundary);
}

// ============================================================================
// Topological Gap -- the cost of closing the loop
// ============================================================================

/**
 * The topological gap: the number of NEW states that appear when
 * a linear mesh is curled into a ring.
 *
 * L_n - F_n = 2·F_{n-1}
 *
 * This is the memory the Betti compiler must pre-allocate when a
 * Swarm edge worker transitions from Open to Closed mid-execution.
 *
 * Proved in Lean: topological_gap_5, topological_gap_6, etc.
 */
export function topologicalGap(n: number): number {
  return lucas(n) - fib(n);
}

/**
 * The gap decomposition: L_n = F_n + 2·F_{n-1}.
 * Returns the two components separately for the allocator.
 */
export function gapDecomposition(n: number): {
  /** State-space of the Open (linear) manifold. */
  readonly openStates: number;
  /** State-space of the Closed (cyclic) manifold. */
  readonly closedStates: number;
  /** Additional states from boundary closure: 2·F_{n-1}. */
  readonly gap: number;
  /** The gap expressed as a multiple of the previous Fibonacci number. */
  readonly gapFactor: 2;
  /** The previous Fibonacci number. */
  readonly fibPrev: number;
} {
  const openStates = fib(n);
  const closedStates = lucas(n);
  const fibPrev = n > 0 ? fib(n - 1) : 0;
  return {
    openStates,
    closedStates,
    gap: closedStates - openStates,
    gapFactor: 2,
    fibPrev,
  };
}

// ============================================================================
// √5 Convergence -- the discriminant ratio
// ============================================================================

/**
 * Compute the convergence ratio L_n / F_n.
 *
 * For large n, this converges to √5 = φ + 1/φ ≈ 2.236.
 * The Pell discriminant proves: (L_n/F_n)² = 5 ± 4/F_n².
 *
 * This is the "Reynolds number" of the Betti manifold: the ratio
 * of closed-state volume to open-state volume.
 */
export function convergenceRatio(n: number): number {
  const fn = fib(n);
  if (fn === 0) return Infinity;
  return lucas(n) / fn;
}

/**
 * Compute the Pell discriminant residual: L_n² - 5·F_n².
 *
 * For even n: L_n² = 5·F_n² + 4  →  residual = +4
 * For odd n:  L_n² = 5·F_n² - 4  →  residual = -4
 *
 * The fact that this is always ±4 (never growing) is the integer
 * proof that L_n/F_n → √5. The "4" is the constant viscosity
 * term that keeps the laminar flow stable.
 *
 * Proved in Lean: pell_discriminant_even_4, pell_discriminant_odd_5, etc.
 */
export function pellDiscriminant(n: number): number {
  const l = lucas(n);
  const f = fib(n);
  return l * l - 5 * f * f;
}

// ============================================================================
// Boundary Transition Safety -- the Betti compiler integration
// ============================================================================

/**
 * Result of a boundary transition analysis.
 * The Betti compiler emits this when a topology edge transitions
 * a manifold from Open to Closed (or vice versa).
 */
export interface BoundaryTransitionResult {
  /** The manifold before the transition. */
  readonly from: IsomorphicManifold;
  /** The manifold after the transition. */
  readonly to: IsomorphicManifold;
  /** State-space before. */
  readonly statesBefore: number;
  /** State-space after. */
  readonly statesAfter: number;
  /** Net new states (positive = growth, negative = contraction). */
  readonly delta: number;
  /** The topological gap at this dimension. */
  readonly gap: number;
  /** Current convergence ratio (L_n/F_n). */
  readonly reynoldsNumber: number;
  /** Is the flow still laminar? (reynolds < φ²) */
  readonly laminar: boolean;
  /** Pell discriminant residual (should be ±4). */
  readonly pellResidual: number;
  /** Is the Pell invariant intact? */
  readonly pellIntact: boolean;
}

/**
 * Analyze a boundary transition.
 *
 * When a Swarm edge worker shifts from an Open path to a Closed ring,
 * the Betti compiler must:
 *   1. Pre-allocate 2·F_{n-1} additional states (the topological gap)
 *   2. Check that the Reynolds number (L_n/F_n) remains below φ²
 *   3. Verify the Pell invariant (L_n² - 5·F_n² = ±4)
 *
 * If the Reynolds number exceeds φ², the laminar execution layers
 * will shear and β₁ collapses — turbulence in the topology.
 */
export function analyzeBoundaryTransition(
  from: IsomorphicManifold,
  to: IsomorphicManifold
): BoundaryTransitionResult {
  const statesBefore = evaluateStateSpace(from);
  const statesAfter = evaluateStateSpace(to);
  const delta = statesAfter - statesBefore;
  const gap = topologicalGap(to.dimension);
  const reynolds = convergenceRatio(to.dimension);
  const pell = pellDiscriminant(to.dimension);

  return {
    from,
    to,
    statesBefore,
    statesAfter,
    delta,
    gap,
    reynoldsNumber: reynolds,
    laminar: reynolds < PHI_SQUARED,
    pellResidual: pell,
    pellIntact: pell === 4 || pell === -4,
  };
}

/**
 * Create an IsomorphicManifold.
 * Factory that mirrors Lean's IsomorphicManifold.mk.
 */
export function createManifold(
  domainName: string,
  dimension: number,
  boundary: BoundaryTopology
): IsomorphicManifold {
  return { domainName, dimension, boundary };
}

/**
 * Transition a manifold from Open to Closed (close the loop).
 * Returns the new manifold and the transition analysis.
 */
export function closeLoop(
  m: IsomorphicManifold
): {
  readonly manifold: IsomorphicManifold;
  readonly transition: BoundaryTransitionResult;
} {
  if (m.boundary === 'Closed') {
    // Already closed — no-op transition.
    const closed = m;
    return {
      manifold: closed,
      transition: analyzeBoundaryTransition(m, closed),
    };
  }
  const closed = createManifold(
    m.domainName + ' [Closed]',
    m.dimension,
    'Closed'
  );
  return {
    manifold: closed,
    transition: analyzeBoundaryTransition(m, closed),
  };
}

/**
 * Transition a manifold from Closed to Open (open the ring).
 * Returns the new manifold and the transition analysis.
 */
export function openRing(
  m: IsomorphicManifold
): {
  readonly manifold: IsomorphicManifold;
  readonly transition: BoundaryTransitionResult;
} {
  if (m.boundary === 'Open') {
    const opened = m;
    return {
      manifold: opened,
      transition: analyzeBoundaryTransition(m, opened),
    };
  }
  const opened = createManifold(
    m.domainName + ' [Opened]',
    m.dimension,
    'Open'
  );
  return {
    manifold: opened,
    transition: analyzeBoundaryTransition(m, opened),
  };
}

// ============================================================================
// Product Doubling -- the circuit identity
// ============================================================================

/**
 * Product doubling: F_n · L_n = F_{2n}.
 *
 * When a circuit closes, the Fibonacci index doubles because
 * current flows through both the Fibonacci and Lucas components.
 * This is the memory cost of a "full duplex" closure.
 *
 * Proved in Lean: product_doubling_4, product_doubling_5, product_doubling_6.
 */
export function productDoubling(n: number): {
  readonly product: number;
  readonly doubleIndex: number;
  readonly equal: boolean;
} {
  const product = fib(n) * lucas(n);
  const doubleIndex = fib(2 * n);
  return { product, doubleIndex, equal: product === doubleIndex };
}

// ============================================================================
// Verification -- runtime checks against Lean proofs
// ============================================================================

/**
 * Verify all critical invariants at a given dimension.
 * This is what the Betti Topology handler calls during compilation
 * to ensure the manifold's structural integrity.
 */
export interface ManifoldVerification {
  readonly dimension: number;
  /** L_n = F_{n-1} + F_{n+1} (the Bragg peak trace). */
  readonly lucasGainHolds: boolean;
  /** L_n = F_n + 2·F_{n-1} (the topological gap). */
  readonly topologicalGapHolds: boolean;
  /** L_n² = 5·F_n² ± 4 (the Pell discriminant). */
  readonly pellDiscriminantHolds: boolean;
  /** F_n · L_n = F_{2n} (the product doubling). */
  readonly productDoublingHolds: boolean;
  /** L_n/F_n converges toward √5 (decreasing error). */
  readonly convergenceHealthy: boolean;
  /** All invariants hold. */
  readonly allHold: boolean;
}

/**
 * Handles the gnosis verify Manifold workflow.
 */
export function verifyManifold(n: number): ManifoldVerification {
  const ln = lucas(n);
  const fn = fib(n);

  // L_n = F_{n-1} + F_{n+1}
  const lucasGainHolds = n >= 1 && ln === fib(n - 1) + fib(n + 1);

  // L_n = F_n + 2·F_{n-1}
  const topologicalGapHolds = n >= 1 && ln === fn + 2 * fib(n - 1);

  // L_n² - 5·F_n² = ±4
  const pell = pellDiscriminant(n);
  const pellDiscriminantHolds = pell === 4 || pell === -4;

  // F_n · L_n = F_{2n}
  const pd = productDoubling(n);
  const productDoublingHolds = pd.equal;

  // Convergence: |L_n/F_n - √5| should be small and decreasing
  const convergenceHealthy =
    fn === 0 || Math.abs(ln / fn - SQRT5) < 1;

  const allHold =
    lucasGainHolds &&
    topologicalGapHolds &&
    pellDiscriminantHolds &&
    productDoublingHolds &&
    convergenceHealthy;

  return {
    dimension: n,
    lucasGainHolds,
    topologicalGapHolds,
    pellDiscriminantHolds,
    productDoublingHolds,
    convergenceHealthy,
    allHold,
  };
}

// ============================================================================
// KAM THEORY: THE DIOPHANTINE CONDITION
//
// Kolmogorov–Arnold–Moser (KAM) theory proves that when a stable
// Hamiltonian system is perturbed by chaos, invariant tori with
// "sufficiently irrational" winding numbers survive. The robustness
// of each torus is determined by how hard its winding number is to
// approximate by rationals.
//
// For ω = φ, the convergents are F_{n+1}/F_n, and the Fibonacci
// Cassini identity proves the approximation error numerator is
// ALWAYS exactly 1 — the minimal possible non-zero value.
//
// The Hurwitz constant √5 = φ + 1/φ is the tight bound: for ALL
// irrationals, |α - p/q| < 1/(√5·q²) has infinitely many
// solutions, and for α = φ, √5 cannot be improved.
// ============================================================================

/**
 * Fibonacci Cassini identity: F_{n-1}·F_{n+1} - F_n² = (-1)^n.
 *
 * The determinant of the 2×2 Fibonacci transfer matrix [[F_{n+1}, F_n],
 * [F_n, F_{n-1}]] is always ±1. This SL(2,ℤ) unimodularity is the
 * mechanism that makes φ the most irrational number: the approximation
 * error when using F_{n+1}/F_n to approximate φ has numerator exactly 1.
 *
 * Returns the Cassini residual: +1 (even n) or -1 (odd n).
 * Any other value indicates corruption in the Fibonacci trace.
 *
 * Proved in Lean: kam_cassini_even_4/6/8/10, kam_cassini_odd_3/5/7/9.
 */
export function fibonacciCassini(n: number): number {
  if (n < 1) return 0;
  return fib(n - 1) * fib(n + 1) - fib(n) * fib(n);
}

/**
 * Phase-space sum of squares: F_n² + F_{n+1}² = F_{2n+1}.
 *
 * The total phase-space volume at the doubled-plus-one index equals
 * the sum of squared amplitudes of the two adjacent Fibonacci modes.
 * This is the Pythagorean structure underlying KAM stability.
 *
 * Proved in Lean: kam_phase_space_3/4/5/6.
 */
export function kamPhaseSpace(n: number): {
  readonly sumOfSquares: number;
  readonly doubledPlusOne: number;
  readonly equal: boolean;
} {
  const sumOfSquares = fib(n) * fib(n) + fib(n + 1) * fib(n + 1);
  const doubledPlusOne = fib(2 * n + 1);
  return { sumOfSquares, doubledPlusOne, equal: sumOfSquares === doubledPlusOne };
}

/**
 * Cross-Fibonacci KAM coupling: F_m·F_n + F_{m-1}·F_{n-1} = F_{m+n-1}.
 *
 * Two resonant Fibonacci modes at indices m and n couple into a
 * single mode at index m+n-1. This is the composition law for KAM
 * torus interactions: golden tori that collide merge into the next
 * torus in the hierarchy instead of destroying each other.
 *
 * Proved in Lean: kam_coupling_3_4, kam_coupling_4_5, etc.
 */
export function kamCoupling(m: number, n: number): {
  readonly coupledProduct: number;
  readonly mergedIndex: number;
  readonly mergedValue: number;
  readonly equal: boolean;
} {
  const coupledProduct = fib(m) * fib(n) + fib(m - 1) * fib(n - 1);
  const mergedValue = fib(m + n - 1);
  return {
    coupledProduct,
    mergedIndex: m + n - 1,
    mergedValue,
    equal: coupledProduct === mergedValue,
  };
}

/**
 * Diophantine resistance metric at dimension n.
 *
 * Measures how well the golden torus at this dimension resists
 * rational resonance. The metric is the scaled approximation
 * error: |φ - F_{n+1}/F_n| · F_n², which converges to 1/√5
 * (the inverse of the Hurwitz constant).
 *
 * Lower values = harder to approximate = more resistant to chaos.
 * The limit 1/√5 ≈ 0.4472 is the universal floor: no irrational
 * number in the universe can be MORE resistant than this.
 */
export function diophantineResistance(n: number): {
  /** The scaled error: |φ - F_{n+1}/F_n| · F_n². */
  readonly scaledError: number;
  /** The Hurwitz limit: 1/√5. */
  readonly hurwitzLimit: number;
  /** How close we are to the limit (should decrease with n). */
  readonly residual: number;
  /** The Cassini determinant (should be ±1). */
  readonly cassiniDet: number;
  /** Is the SL(2,ℤ) armor intact? */
  readonly armorIntact: boolean;
} {
  const fn = fib(n);
  const fnp1 = fib(n + 1);
  const hurwitzLimit = 1 / SQRT5;

  if (fn === 0) {
    return {
      scaledError: Infinity,
      hurwitzLimit,
      residual: Infinity,
      cassiniDet: 0,
      armorIntact: false,
    };
  }

  const ratio = fnp1 / fn;
  const error = Math.abs(PHI - ratio);
  const scaledError = error * fn * fn;
  const cassiniDet = fibonacciCassini(n);

  return {
    scaledError,
    hurwitzLimit,
    residual: Math.abs(scaledError - hurwitzLimit),
    cassiniDet,
    armorIntact: cassiniDet === 1 || cassiniDet === -1,
  };
}

/**
 * KAM survival predicate.
 *
 * Returns true if a manifold at this dimension has sufficient
 * Diophantine armor to survive chaotic perturbation. Checks:
 *
 * 1. Cassini determinant is ±1 (SL(2,ℤ) unimodularity)
 * 2. Pell discriminant is ±4 (√5 convergence)
 * 3. Phase-space sum-of-squares identity holds
 * 4. Convergence ratio is below the critical Reynolds number (φ²)
 *
 * If all hold, the golden torus at this dimension is immune
 * to rational resonance cascade per KAM theory.
 */
export function kamSurvival(n: number): {
  readonly dimension: number;
  readonly cassiniIntact: boolean;
  readonly pellIntact: boolean;
  readonly phaseSpaceIntact: boolean;
  readonly laminar: boolean;
  readonly immune: boolean;
} {
  const cassiniDet = fibonacciCassini(n);
  const cassiniIntact = cassiniDet === 1 || cassiniDet === -1;
  const pell = pellDiscriminant(n);
  const pellIntact = pell === 4 || pell === -4;
  const ps = kamPhaseSpace(n);
  const phaseSpaceIntact = ps.equal;
  const reynolds = convergenceRatio(n);
  const laminar = reynolds < PHI_SQUARED;

  return {
    dimension: n,
    cassiniIntact,
    pellIntact,
    phaseSpaceIntact,
    laminar,
    immune: cassiniIntact && pellIntact && phaseSpaceIntact && laminar,
  };
}
