// @a0n/gnosis/src/betti/build-config.ts
/**
 * build-config.ts -- Polyglot race configuration for Betti.
 *
 * Each betti.gg node becomes a raceable function entry, and the polyglot
 * build system races TypeScript/Rust/Go/Python implementations of each
 * handler to find the fastest for compile-time optimization.
 */

import * as fs from 'fs';
import * as path from 'path';
import { BettyCompiler, type GraphAST } from '../betty/compiler';
import type { PolyglotBuildConfig } from '../polyglot-build';
import {
  buildWithTopoRace,
  type PolyglotBuildResult,
} from '../polyglot-build';
import {
  scaffoldTopology,
  type ScaffoldAssignment,
  type ScaffoldResult,
} from '../polyglot-scaffold';
import type { ExtractedFunction } from '../polyglot-compose';

// ============================================================================
// Config
// ============================================================================

/**
 * Default Betti polyglot build config.
 * Source is betti.gg, candidate languages are the core four.
 */
export function createBettiBuildConfig(
  bettiSourcePath?: string
): PolyglotBuildConfig {
  const sourcePath =
    bettiSourcePath ?? path.resolve(__dirname, '../../betti.gg');

  const bettiSource = fs.readFileSync(sourcePath, 'utf-8');

  return {
    sourceFiles: [sourcePath],
    candidateLanguages: ['typescript', 'rust', 'go', 'python'],
    testInput: bettiSource,
    maxCandidates: 4,
    raceTimeoutMs: 10000,
    speculative: true,
    skipErrorThreshold: 0.5,
  };
}

/**
 * Betti handler names that are raceable.
 * Each maps to a function that can be scaffolded in multiple languages.
 */
export const BETTI_RACEABLE_HANDLERS = [
  'source_reader',
  'strip_comments',
  'node_lexer',
  'edge_lexer',
  'property_lexer',
  'ast_assembler',
  'betti_verifier',
  'wasm_emitter',
  'vhdl_emitter',
] as const;

export type BettiHandler = (typeof BETTI_RACEABLE_HANDLERS)[number];

// ============================================================================
// Scaffold
// ============================================================================

/**
 * Generate polyglot skeletons for all betti.gg handler nodes.
 * Uses scaffoldTopology to create Rust/Go/Python/TypeScript implementations.
 */
export function scaffoldBettiHandlers(
  languages: string[] = ['typescript', 'rust', 'go', 'python'],
  bettiSourcePath?: string,
  outputDir?: string
): ScaffoldResult {
  const sourcePath =
    bettiSourcePath ?? path.resolve(__dirname, '../../betti.gg');
  const source = fs.readFileSync(sourcePath, 'utf-8');

  const compiler = new BettyCompiler();
  const result = compiler.parse(source);
  if (!result.ast) {
    throw new Error('Failed to parse betti.gg');
  }

  // Create scaffold assignments for each handler node in each language
  const assignments: ScaffoldAssignment[] = [];
  for (const [nodeId, node] of result.ast.nodes) {
    // Only scaffold nodes that correspond to betti handlers (have a label)
    if (node.labels.length === 0) continue;
    for (const lang of languages) {
      assignments.push({
        nodeId,
        language: lang,
        functionName: nodeId,
      });
    }
  }

  const scaffolded = scaffoldTopology(result.ast, assignments, source);

  // Write files if outputDir provided
  if (outputDir) {
    fs.mkdirSync(outputDir, { recursive: true });
    for (const file of scaffolded.files) {
      const filePath = path.join(outputDir, file.fileName);
      fs.writeFileSync(filePath, file.source, 'utf8');
    }
  }

  return scaffolded;
}

// ============================================================================
// Build
// ============================================================================

/**
 * Convert betti.gg handler nodes into ExtractedFunction[] for the polyglot build system.
 */
export function extractBettiFunctions(
  bettiSourcePath?: string
): ExtractedFunction[] {
  const sourcePath =
    bettiSourcePath ?? path.resolve(__dirname, '../../betti.gg');
  const source = fs.readFileSync(sourcePath, 'utf-8');

  const compiler = new BettyCompiler();
  const result = compiler.parse(source);
  if (!result.ast) {
    throw new Error('Failed to parse betti.gg');
  }

  const functions: ExtractedFunction[] = [];
  for (const [nodeId, node] of result.ast.nodes) {
    if (node.labels.length === 0) continue;

    // Find edges involving this node to determine edge types
    const edgeTypes = new Set<string>();
    const callees: string[] = [];
    for (const edge of result.ast.edges) {
      if (edge.sourceIds.includes(nodeId)) {
        edgeTypes.add(edge.type);
        callees.push(...edge.targetIds);
      }
    }

    // Build a mini-topology for this handler
    const miniNodes = new Map<string, typeof node>();
    miniNodes.set(nodeId, node);
    const miniEdges = result.ast.edges.filter(
      (e) => e.sourceIds.includes(nodeId) || e.targetIds.includes(nodeId)
    );

    functions.push({
      filePath: sourcePath,
      language: 'typescript',
      name: nodeId,
      params: ['payload', 'props'],
      callees,
      ast: { nodes: miniNodes, edges: miniEdges },
      ggSource: `(${nodeId}:${node.labels[0]}${
        Object.keys(node.properties).length > 0
          ? ` { ${Object.entries(node.properties)
              .map(([k, v]) => `${k}: '${v}'`)
              .join(', ')} }`
          : ''
      })`,
      nodeCount: 1,
      edgeTypes: [...edgeTypes],
    });
  }

  return functions;
}

/**
 * Run a polyglot build race for all betti.gg handlers.
 * Races TypeScript/Rust/Go/Python implementations, records winners in strategy memory.
 */
export async function buildBetti(
  bettiSourcePath?: string,
  config?: Partial<PolyglotBuildConfig>
): Promise<PolyglotBuildResult> {
  const functions = extractBettiFunctions(bettiSourcePath);
  const buildConfig = createBettiBuildConfig(bettiSourcePath);

  return buildWithTopoRace(functions, {
    ...buildConfig,
    ...config,
  });
}
