// @a0n/gnosis/src/betti/buley-equilibrium.ts
/**
 * buley-equilibrium.ts — TypeScript mirror of BuleyEquilibrium.lean.
 *
 * Mirrors:
 *   - BuleyState (manifold_vector + temporal_debt)
 *   - isBuleyEquilibrium (Novikov + Algebraic)
 *   - isAbsoluteZero (the a0 Trinity)
 *   - buley_equilibrium_existence (at Decimal Fixed Point n=10)
 *   - buley_refines_skyrms
 *   - absolute_zero_reachability
 *   - buley_ssm_convergence
 */

import { fib } from './boundary-topology';
import type { VectorState } from './topological-memoization';
import {
  type TopologicalDebt,
  createTopologicalDebt,
  satisfiesNovikov,
} from './retrocausal-memoization';
import { isAlgebraicEquilibrium } from './skyrms-syzygy-equilibrium';

// ============================================================================
// BuleyState
// ============================================================================

/**
 * Buley Equilibrium State (Θ_B).
 *
 * A higher-order signaling equilibrium where the convention is enforced
 * retrocausally across all geometric regimes (Linear, Triangular,
 * Classical, Quantum).
 *
 * Mirrors BuleyEquilibrium.BuleyState.
 */
export interface BuleyState {
  /** The Moonshine vector (bulkState maps to Fibonacci execution). */
  readonly manifoldVector: { readonly bulkState: number };
  /** The retrocausal debt linking present to future. */
  readonly temporalDebt: TopologicalDebt;
}

/**
 * Create a BuleyState.
 */
export function createBuleyState(
  bulkState: number,
  futureN: number,
  timestamp: number
): BuleyState {
  return {
    manifoldVector: { bulkState },
    temporalDebt: createTopologicalDebt({ n: futureN }, timestamp),
  };
}

// ============================================================================
// Buley Invariant
// ============================================================================

/**
 * The Buley Invariant:
 * 1. Temporal: Satisfies the Novikov Self-Consistency Invariant.
 * 2. Algebraic: Sits at a Skyrms-Algebraic Syzygy (The Full Eclipse).
 *
 * Mirrors BuleyEquilibrium.isBuleyEquilibrium.
 */
export function isBuleyEquilibrium(s: BuleyState): boolean {
  // Novikov: bulkState dimension matches future_output dimension
  const novikovActual: VectorState = { n: s.manifoldVector.bulkState };
  const novikovHolds = satisfiesNovikov(novikovActual, s.temporalDebt);

  // Algebraic: bulkState sits at a Full Eclipse
  const algebraicHolds = isAlgebraicEquilibrium({
    v: s.manifoldVector,
    isInvariant: true,
  });

  return novikovHolds && algebraicHolds;
}

// ============================================================================
// Absolute Zero (a0)
// ============================================================================

/**
 * The Absolute Zero Trinity (a0):
 * The { Nash, Skyrms, Buley } trifecta of stabilization.
 *
 * At Absolute Zero, topological friction (β₁) is 0.
 * The execution is no longer a process; it is a static, autonomic
 * geometric fact. This is the Goodchild Gate.
 *
 * Mirrors BuleyEquilibrium.isAbsoluteZero.
 */
export function isAbsoluteZero(s: BuleyState): boolean {
  return isBuleyEquilibrium(s);
}

// ============================================================================
// Verification
// ============================================================================

export interface BuleyExistenceVerification {
  readonly state: BuleyState;
  readonly novikovHolds: boolean;
  readonly algebraicHolds: boolean;
  readonly equilibrium: boolean;
  /** Proved in Lean: buley_equilibrium_existence. */
  readonly verified: boolean;
}

/**
 * Verify the existence of a Buley Equilibrium at the Decimal Fixed Point.
 *
 * At n=10: F₁₀ = 55, T₁₀ = 55. The bulkState 55 is both Fibonacci-indexed
 * and Triangular, satisfying the algebraic eclipse. The Novikov constraint
 * is satisfied by matching future_output.n to the bulkState dimension.
 *
 * Proved in Lean: buley_equilibrium_existence.
 */
export function verifyBuleyEquilibriumExistence(): BuleyExistenceVerification {
  // The Decimal Fixed Point: bulkState = 55, future_output.n = 55
  const s = createBuleyState(55, 55, 0);
  const novikovActual: VectorState = { n: 55 };
  const novikovHolds = satisfiesNovikov(novikovActual, s.temporalDebt);
  const algebraicHolds = isAlgebraicEquilibrium({
    v: s.manifoldVector,
    isInvariant: true,
  });
  const eq = isBuleyEquilibrium(s);

  return {
    state: s,
    novikovHolds,
    algebraicHolds,
    equilibrium: eq,
    verified: eq,
  };
}

export interface BuleyRefinesSkyrmsVerification {
  readonly state: BuleyState;
  readonly buleyHolds: boolean;
  readonly skyrmsHolds: boolean;
  /** Proved in Lean: buley_refines_skyrms. */
  readonly verified: boolean;
}

/**
 * Verify that Buley strictly contains Skyrms.
 * If a state is a Buley Equilibrium, it is also an Algebraic Equilibrium.
 *
 * Proved in Lean: buley_refines_skyrms.
 */
export function verifyBuleyRefinesSkyrms(): BuleyRefinesSkyrmsVerification {
  const s = createBuleyState(55, 55, 0);
  const buleyHolds = isBuleyEquilibrium(s);
  const skyrmsHolds = isAlgebraicEquilibrium({
    v: s.manifoldVector,
    isInvariant: true,
  });
  return {
    state: s,
    buleyHolds,
    skyrmsHolds,
    // Theorem: buley → skyrms (if buley holds, skyrms must also hold)
    verified: !buleyHolds || skyrmsHolds,
  };
}

export interface AbsoluteZeroVerification {
  readonly state: BuleyState;
  readonly isAbsoluteZero: boolean;
  /** Proved in Lean: absolute_zero_reachability. */
  readonly verified: boolean;
}

/**
 * Verify Absolute Zero reachability.
 *
 * Proved in Lean: absolute_zero_reachability.
 */
export function verifyAbsoluteZeroReachability(): AbsoluteZeroVerification {
  const s = createBuleyState(55, 55, 0);
  const az = isAbsoluteZero(s);
  return { state: s, isAbsoluteZero: az, verified: az };
}

/** Full verification suite. */
export function verifyBuleyEquilibrium(): {
  readonly existence: BuleyExistenceVerification;
  readonly refinesSkyrms: BuleyRefinesSkyrmsVerification;
  readonly absoluteZero: AbsoluteZeroVerification;
  readonly allHold: boolean;
} {
  const existence = verifyBuleyEquilibriumExistence();
  const refinesSkyrms = verifyBuleyRefinesSkyrms();
  const absoluteZero = verifyAbsoluteZeroReachability();
  return {
    existence,
    refinesSkyrms,
    absoluteZero,
    allHold: existence.verified && refinesSkyrms.verified && absoluteZero.verified,
  };
}
