// @a0n/gnosis/src/betti/pisot-primitive.ts
/**
 * pisot-primitive.ts -- Pisot Vickrey CPU as a Gnosis FlowFrame primitive.
 *
 * Mirrors open-source/gnosis/lean/Lean/ForkRaceFoldTheorems/PisotVickreyTable.lean
 * and ZeckendorfFST.lean. The Lean files are the spec; this module is the
 * TypeScript runtime that .gg topologies call into.
 *
 * Opcodes:
 *   fork   -- push F[n] onto the register stack (fork into a Fibonacci slot)
 *   race   -- pop two slots and keep the winner via Vickrey second-price rule
 *   smooth -- Zeckendorf-smooth a natural number into non-adjacent Fib indices
 *   check  -- run the Cassini invariant: F[n-1]*F[n+1] - F[n]^2 == (-1)^n
 *
 * The CHECK invariant runs implicitly after every FORK/RACE. If it fails the
 * runtime throws, which halts the enclosing topology -- the TS mirror of
 * `cassini_no_silent_corruption` from PisotVickreyTable.lean.
 */

import type { GnosisRegistry } from '../runtime/registry.js';

// ============================================================================
// Fibonacci table (matches Lean F: F[0] = F[1] = 1, F[n+2] = F[n+1] + F[n])
// ============================================================================

/** Cap at 46 so F[46] = 2971215073 stays within u32 safety. */
export const PISOT_MAX_INDEX = 46;

export const PISOT_FIB: ReadonlyArray<number> = (() => {
  const f = new Array<number>(PISOT_MAX_INDEX + 1);
  f[0] = 1;
  f[1] = 1;
  for (let n = 2; n <= PISOT_MAX_INDEX; n++) {
    const prev = f[n - 1];
    const prevPrev = f[n - 2];
    if (prev === undefined || prevPrev === undefined) {
      throw new Error(`[Pisot] fib table seed missing at index ${n}`);
    }
    f[n] = prev + prevPrev;
  }
  return f;
})();

/** Fibonacci lookup with bounds check (matches Lean F). */
export function pisotF(n: number): number {
  if (!Number.isInteger(n) || n < 0 || n > PISOT_MAX_INDEX) {
    throw new Error(
      `[Pisot] index ${n} out of range [0, ${PISOT_MAX_INDEX}]`
    );
  }
  const value = PISOT_FIB[n];
  if (value === undefined) {
    throw new Error(`[Pisot] fib table missing value at index ${n}`);
  }
  return value;
}

// ============================================================================
// Cassini invariant
// ============================================================================

/**
 * Cassini parity: F[n-1] * F[n+1] - F[n]^2 == (-1)^n.
 * Mirrors `cassini_no_silent_corruption` in PisotVickreyTable.lean.
 */
export function pisotCassini(n: number): boolean {
  if (n < 1 || n + 1 > PISOT_MAX_INDEX) return true; // vacuous at boundaries
  // Use BigInt so F[n-1]*F[n+1] stays exact at the high end of the table
  // (products exceed 2^53 beyond n ~ 27).
  const a = BigInt(pisotF(n - 1));
  const b = BigInt(pisotF(n + 1));
  const c = BigInt(pisotF(n));
  const lhs = a * b - c * c;
  // Note: our table is shifted (F[0]=F[1]=1 vs classical F[0]=0,F[1]=1), so
  // the Cassini sign is flipped relative to the classical statement.
  const rhs = n % 2 === 0 ? -1n : 1n;
  return lhs === rhs;
}

/** Throw-if-broken CHECK. */
export function pisotCheck(n: number): void {
  if (!pisotCassini(n)) {
    throw new Error(
      `[Pisot:CHECK] Cassini parity violated at index ${n} -- halting topology`
    );
  }
}

// ============================================================================
// Register stack CPU
// ============================================================================

export interface PisotState {
  /** Stack of Fibonacci slot indices (each slot holds F[index]). */
  stack: number[];
  /** Last opcode executed (for trace). */
  lastOp: PisotOp | null;
}

export type PisotOp = 'fork' | 'race' | 'smooth' | 'check';

export interface PisotInstruction {
  op: PisotOp;
  /** Operand: slot index for fork, ignored for race, natural number for smooth, index for check. */
  arg?: number;
}

/**
 * Handles the gnosis pisot Init workflow.
 */
export function pisotInit(): PisotState {
  return { stack: [], lastOp: null };
}

/**
 * FORK: push a Fibonacci slot index onto the stack. Post-check Cassini.
 * Mirrors Lean `fork` on the Pisot Vickrey table.
 */
export function pisotFork(state: PisotState, index: number): PisotState {
  pisotF(index); // bounds check
  const next = { stack: [...state.stack, index], lastOp: 'fork' as const };
  pisotCheck(index);
  return next;
}

/**
 * RACE: pop two slots, keep the loser's bid at the winner's slot (Vickrey
 * second-price). Here we keep max(index) as the winner and drop the other --
 * matches Lean `race` which selects the higher Fibonacci weight.
 */
export function pisotRace(state: PisotState): PisotState {
  if (state.stack.length < 2) {
    throw new Error(
      `[Pisot:RACE] need 2 slots, have ${state.stack.length}`
    );
  }
  const stack = [...state.stack];
  const b = stack.pop()!;
  const a = stack.pop()!;
  const winner = Math.max(a, b);
  stack.push(winner);
  pisotCheck(winner);
  return { stack, lastOp: 'race' };
}

/**
 * SMOOTH: Zeckendorf decomposition of `m` into non-adjacent Fibonacci indices.
 * Mirrors `smooth`/`dec`/`nonAdjacent` in ZeckendorfFST.lean.
 * Returns descending list of indices i such that sum F[i] = m and no two
 * indices are adjacent.
 */
export function pisotSmooth(m: number): number[] {
  if (!Number.isInteger(m) || m < 0) {
    throw new Error(`[Pisot:SMOOTH] input ${m} not a natural number`);
  }
  if (m === 0) return [];
  const out: number[] = [];
  let remaining = m;
  // Walk down from the largest usable Fib index.
  for (let i = PISOT_MAX_INDEX; i >= 1 && remaining > 0; i--) {
    const fi = PISOT_FIB[i];
    if (fi === undefined) {
      continue;
    }
    if (fi <= remaining) {
      out.push(i);
      remaining -= fi;
      i--; // enforce non-adjacency: skip the next index
    }
  }
  if (remaining > 0) {
    throw new Error(`[Pisot:SMOOTH] residue ${remaining} -- m too large`);
  }
  return out;
}

/** Execute one instruction against a state, returning the next state. */
export function pisotStep(
  state: PisotState,
  instr: PisotInstruction
): PisotState {
  switch (instr.op) {
    case 'fork':
      return pisotFork(state, instr.arg ?? 0);
    case 'race':
      return pisotRace(state);
    case 'smooth': {
      const indices = pisotSmooth(instr.arg ?? 0);
      return {
        stack: [...state.stack, ...indices],
        lastOp: 'smooth',
      };
    }
    case 'check':
      pisotCheck(instr.arg ?? state.stack[state.stack.length - 1] ?? 0);
      return { ...state, lastOp: 'check' };
    default:
      throw new Error(`[Pisot] unknown op: ${(instr as PisotInstruction).op}`);
  }
}

/** Run a full program, returning the final state. */
export function pisotRun(program: PisotInstruction[]): PisotState {
  let state = pisotInit();
  for (const instr of program) {
    state = pisotStep(state, instr);
  }
  return state;
}

// ============================================================================
// Betti emission: compile pisot-labeled AST nodes to a bytecode sequence
// ============================================================================

export interface PisotBytecode {
  kind: 'pisot-bytecode';
  instructions: PisotInstruction[];
}

/**
 * Scan a GraphAST-like payload for nodes labeled `Pisot` (or `PisotVickrey`)
 * and compile them into a linear bytecode stream for the emitter.
 *
 * Node properties expected:
 *   { op: 'fork'|'race'|'smooth'|'check', arg?: '<number>' }
 */
export function compilePisotNodes(ast: {
  nodes: Map<string, { labels: string[]; properties: Record<string, string> }>;
}): PisotBytecode {
  const instructions: PisotInstruction[] = [];
  for (const [, node] of ast.nodes) {
    const isPisot = node.labels.some(
      (l) => l === 'Pisot' || l === 'PisotVickrey'
    );
    if (!isPisot) continue;
    const op = (node.properties['op'] ?? '').toLowerCase() as PisotOp;
    if (!['fork', 'race', 'smooth', 'check'].includes(op)) {
      throw new Error(`[Pisot:compile] invalid op '${op}' on node`);
    }
    const argRaw = node.properties['arg'];
    const arg = argRaw !== undefined ? Number(argRaw) : undefined;
    instructions.push(arg === undefined ? { op } : { op, arg });
  }
  return { kind: 'pisot-bytecode', instructions };
}

// ============================================================================
// Registration
// ============================================================================

/**
 * Register the Pisot primitive handler. Topologies declare
 *   (node: Pisot { op: 'fork', arg: '5' })
 * and invoke via the normal handler dispatch.
 *
 * The handler now integrates KAM survival checks from boundary-topology.ts:
 * after every FORK/RACE, it verifies the Cassini determinant and reports
 * the armor status in the returned state.
 */
export function registerPisotPrimitive(registry: GnosisRegistry): void {
  const handler = async (
    payload: any,
    props: Record<string, string>
  ): Promise<any> => {
    const op = (props['op'] ?? '').toLowerCase() as PisotOp;
    const arg = props['arg'] !== undefined ? Number(props['arg']) : undefined;
    const state: PisotState =
      payload && payload.kind === 'pisot-state' ? payload : pisotInit();
    const next = pisotStep(state, arg === undefined ? { op } : { op, arg });

    // KAM armor telemetry: verify Cassini at the top of stack
    const topIndex = next.stack[next.stack.length - 1];
    let kamStatus: 'immune' | 'vulnerable' | 'n/a' = 'n/a';
    if (topIndex !== undefined && topIndex >= 2) {
      const cassini = pisotCassini(topIndex);
      kamStatus = cassini ? 'immune' : 'vulnerable';
    }

    console.log(
      `[Pisot:${op}] stack=[${next.stack.join(',')}] kam=${kamStatus}`
    );
    return { kind: 'pisot-state', ...next, kamStatus };
  };
  registry.register('Pisot', handler, { override: true });
  registry.register('PisotVickrey', handler, { override: true });
}
