// @a0n/.ignored_gnosis/.workspace-stubs/aeon-logic/probability-routing.js
const DEFAULT_MIN_OBSERVATIONS = 3;
const DEFAULT_MIN_CONCENTRATION_BPS = 5000;
const DEFAULT_MAX_ROUTES = 5;

function assertRoute(route) {
  if (typeof route !== 'string' || route.trim().length === 0) {
    throw new Error('route must be a non-empty string');
  }
}

function assertNat(value, label) {
  if (!Number.isSafeInteger(value) || value < 0) {
    throw new Error(`${label} must be a safe natural number`);
  }
}

function probabilityRatio(numerator, denominator) {
  return { numerator, denominator };
}

function routeDecision(route, count, totalMass, policy) {
  const residualMass = totalMass - count;
  const concentrationBps = Math.floor((count * 10000) / totalMass);

  if (totalMass < policy.minObservations) {
    return {
      decision: 'explore',
      reason: `${route} has insufficient observed mass for reuse.`,
    };
  }

  if (residualMass > policy.residualBudget) {
    return {
      decision: 'cover',
      reason: `${route} leaves residual route mass above the configured cover budget.`,
    };
  }

  if (concentrationBps >= policy.minConcentrationBps) {
    return {
      decision: 'reuse',
      reason: `${route} concentrates enough observed mass for exact reuse.`,
    };
  }

  return {
    decision: 'speculate',
    reason: `${route} is covered but not concentrated enough for exact reuse.`,
  };
}

/**
 * Handles the gnosis select Probability Routes workflow.
 */
export function selectProbabilityRoutes(masses, options = {}) {
  const normalized = masses
    .map((mass, index) => {
      assertRoute(mass.route);
      assertNat(mass.count, `masses[${index}].count`);
      return { route: mass.route, count: mass.count };
    })
    .filter((mass) => mass.count > 0);
  const totalMass = normalized.reduce((sum, mass) => sum + mass.count, 0);
  const policy = {
    minObservations: options.minObservations ?? DEFAULT_MIN_OBSERVATIONS,
    minConcentrationBps:
      options.minConcentrationBps ?? DEFAULT_MIN_CONCENTRATION_BPS,
    residualBudget: options.residualBudget ?? totalMass,
    maxRoutes: options.maxRoutes ?? DEFAULT_MAX_ROUTES,
  };

  assertNat(policy.minObservations, 'minObservations');
  assertNat(policy.minConcentrationBps, 'minConcentrationBps');
  assertNat(policy.residualBudget, 'residualBudget');
  assertNat(policy.maxRoutes, 'maxRoutes');
  if (policy.minConcentrationBps > 10000) {
    throw new Error('minConcentrationBps cannot exceed 10000');
  }
  if (policy.maxRoutes === 0) {
    throw new Error('maxRoutes must be positive');
  }

  const witnesses = normalized
    .map((mass) => {
      const decision = routeDecision(mass.route, mass.count, totalMass, policy);
      return {
        route: mass.route,
        count: mass.count,
        totalMass,
        residualMass: totalMass - mass.count,
        concentrationBps: Math.floor((mass.count * 10000) / totalMass),
        probability: probabilityRatio(mass.count, totalMass),
        decision: decision.decision,
        reason: decision.reason,
      };
    })
    .sort(
      (left, right) =>
        right.count - left.count ||
        right.concentrationBps - left.concentrationBps ||
        left.route.localeCompare(right.route)
    );

  const selected = witnesses
    .filter(
      (witness) =>
        witness.decision === 'reuse' || witness.decision === 'speculate'
    )
    .slice(0, policy.maxRoutes);

  return {
    totalMass,
    witnesses,
    selected,
  };
}
