// @a0n/.ignored_gnosis/forkjoin-app/client/hooks/useCursorTracking.ts
import { useEffect, useRef, useCallback, useState } from '@a0n/raect';
import type { CursorViewportContext } from './usePresence';

interface CursorPosition {
  x: number;
  y: number;
}

interface UseCursorTrackingOptions {
  sendCursor: (x: number, y: number, context?: CursorViewportContext) => void;
  enabled?: boolean;
}

interface UseCursorTrackingResult {
  position: CursorPosition;
}

/**
 * Provides the use Cursor Tracking hook.
 */
export function useCursorTracking({
  sendCursor,
  enabled = true,
}: UseCursorTrackingOptions): UseCursorTrackingResult {
  const [position, setPosition] = useState<CursorPosition>({ x: 0, y: 0 });
  const lastSentRef = useRef<number>(0);

  // Throttle to 30hz
  const THROTTLE_MS = 33;

  const handleMouseMove = useCallback(
    (event: MouseEvent) => {
      const now = Date.now();

      // Update local position immediately (for smooth local cursor if needed)
      const viewportWidth = Math.max(window.innerWidth, 1);
      const viewportHeight = Math.max(window.innerHeight, 1);
      const x = event.clientX / viewportWidth;
      const y = event.clientY / viewportHeight;
      setPosition({ x, y });

      // Throttle network sends
      if (now - lastSentRef.current < THROTTLE_MS) return;
      lastSentRef.current = now;

      // Send normalized position (0-1)
      sendCursor(x, y, {
        scrollX: window.scrollX,
        scrollY: window.scrollY,
        viewportWidth,
        viewportHeight,
      });
    },
    [sendCursor]
  );

  useEffect(() => {
    if (!enabled) return;
    window.addEventListener('mousemove', handleMouseMove, { passive: true });
    return () => {
      window.removeEventListener('mousemove', handleMouseMove);
    };
  }, [enabled, handleMouseMove]);

  return { position };
}
