// @a0n/gnosis/gnode/benchmarks/cross-language-bench.mjs
#!/usr/bin/env node
/* global console, process */
/**
 * Cross-language Monster benchmark harness.
 *
 * Walks every known orchestrator-shaped fixture (TS + Python + future
 * Go/Rust/Java), runs each through the prebuilt `monster` binary 5x,
 * captures wall time / pass-fail / output, and emits both a structured
 * JSON dump and a Markdown summary.
 *
 * Designed as:
 *   1. A regression safety net (diff two runs of the JSON to detect drift).
 *   2. A perf table source-of-truth for the writeup.
 *
 * Constraints (locked in by the writeup spec):
 *   - Pure Node builtins, no new deps.
 *   - Does NOT rebuild polyglot. Uses `polyglot/target/release/monster` as-is.
 *   - Skips fixtures that don't exist (other agents are mid-flight).
 *   - Exit 0 if all known-good cases pass; non-zero on regression.
 */

import { spawnSync } from 'node:child_process';
import { existsSync, statSync, writeFileSync } from 'node:fs';
import path from 'node:path';
import { performance } from 'node:perf_hooks';
import { fileURLToPath } from 'node:url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const gnodeDir = path.resolve(__dirname, '..');
const gnosisDir = path.resolve(gnodeDir, '..');
const repoRoot = path.resolve(gnosisDir, '../..');

const GNEXEC_TMP = '/tmp/gnexec-target/release/gnexec';
const MONSTER_TMP = '/tmp/gnexec-target/release/monster';
const GNOSIS_POLYGLOT_TMP = '/tmp/gnexec-target/release/gnosis-polyglot';

const MONSTER_LOCAL = path.resolve(gnosisDir, 'polyglot/target/release/monster');
const GNEXEC_LOCAL = path.resolve(gnosisDir, 'polyglot/target/release/gnexec');
const GNEXEC = existsSync(MONSTER_TMP)
  ? MONSTER_TMP
  : existsSync(MONSTER_LOCAL)
    ? MONSTER_LOCAL
    : existsSync(GNEXEC_TMP)
      ? GNEXEC_TMP
      : GNEXEC_LOCAL;
const MONSTER_RUNTIME_NAME = path.basename(GNEXEC) === 'monster' ? 'monster' : 'gnexec';
const GNOSIS_POLYGLOT = existsSync(GNOSIS_POLYGLOT_TMP)
  ? GNOSIS_POLYGLOT_TMP
  : path.resolve(gnosisDir, 'polyglot/target/release/gnosis-polyglot');
const POLYGLOT_LIB = path.resolve(
  gnosisDir,
  'polyglot/target/release/libgnosis_polyglot.rlib'
);
const BECKY = path.resolve(gnosisDir, 'polyglot/target/release/becky');
const INVOKE_APP = path.resolve(__dirname, 'invoke-app.mjs');
const RESULTS_JSON = '/tmp/gnexec-bench-results.json';

const ITERATIONS = 5; // 1 warm + 4 measured
const TIMEOUT_MS = 30_000;

/**
 * Per-language wall-time budgets (median, ms). Run fails loudly if a
 * language median crosses this ceiling — wall-time guardrail, not exit
 * code. Only applied to gnexec-runtime `pass`-kind cases (extractor IR
 * cases are excluded — they run a different binary).
 *
 * Note: today's go/rust/java runtime medians are computed from a single
 * `<lang>-echo` case (kind=expected-unsupported, exits fast). On a noisy
 * laptop the single-sample median can spike past 25 ms purely from
 * scheduler noise. Once gnexec consumes structured IR and these cases
 * become real `pass` cases, samples stabilize. Until then, this budget
 * is a regression *detector* — a steady-state >25 ms median is real;
 * one-off spikes are noise. Re-run on failure before treating as a
 * regression.
 */
const LANGUAGE_BUDGET_MS = {
  ts: 25,
  py: 25,
  go: 25,
  rust: 25,
  java: 25,
};

/**
 * Per-case slow-regression ceilings. Map of label → ms. Hard-fail if a
 * specific case exceeds this ceiling even when its language median is OK.
 * Currently used to lock in the post-#32 (per-Ctx scratch-arena) target
 * for try/catch — current baseline is ~28 ms, target is ~24 ms.
 */
const CASE_SLOW_CEILING_MS = {
  // OPTIMIZATION TARGET (#32 per-Ctx scratch-arena): def-trycatch is the
  // slowest case in the suite (~28 ms steady). Once the per-Ctx scratch
  // arena lands, expect convergence to within 2× of baseline (~24 ms).
  // Set ceiling at 50 ms — fast-fail on a *worse* regression, not on the
  // current baseline. Tighten to 25 ms after #32 ships.
  'def-trycatch': 50,
};

/* -------------------------------------------------------------------------- */
/* Test case registry                                                         */
/* -------------------------------------------------------------------------- */

/**
 * Each case: { file, input, expect, label, lang, exportName?, kind? }
 *   - expect: string for exact match, RegExp for predicate, or function(stdout)=>bool.
 *   - kind: "pass" (default) | "expected-throw" (exit 1, message match)
 *           | "expected-unsupported" (exit 4 unless `expectedExit` overrides)
 *           | "extractor-emits-ir" — runs gnosis-polyglot (NOT gnexec) and
 *              greps the stdout for IR markers (expr=, bind=, callee=,
 *              args=). This locks in the IR-emitter regression net,
 *              independent of whether gnexec evaluates the structured IR.
 *           Treat all "known" categories as regressions only on *change*,
 *           not on current state.
 */
const cases = [
  /* ─── Canonical TS baselines (also used for JS-bridge floor) ──────── */
  {
    label: 'echo',
    file: path.resolve(__dirname, 'echo.ts'),
    input: { name: 'gnode' },
    expect: 'hello:gnode',
    lang: 'ts',
    floor: true,
  },
  {
    label: 'fib',
    file: path.resolve(__dirname, 'fib.ts'),
    input: { n: 10 },
    expect: '55',
    lang: 'ts',
    floor: true,
  },
  {
    label: 'fanout',
    file: path.resolve(__dirname, 'fanout.ts'),
    input: { userId: 'u1' },
    expect: 'u1:user|u1:profile',
    lang: 'ts',
    floor: true,
  },

  /* ─── Path A: Python landed (echo, fib) ───────────────────────────── */
  {
    label: 'py-echo',
    file: '/tmp/gnexec-py/echo.py',
    input: { name: 'world' },
    expect: 'hello:world',
    lang: 'py',
  },
  {
    label: 'py-fib',
    file: '/tmp/gnexec-py/fib.py',
    input: { n: 10 },
    expect: '55',
    lang: 'py',
  },

  /* ─── Coverage round 1 (cov fixtures, all use export `app`) ───────── */
  {
    label: 'cov-length',
    file: '/tmp/gnexec-cov/t-length.ts',
    input: { xs: [1, 2, 3] },
    expect: '3',
    lang: 'ts',
  },
  {
    label: 'cov-toupper',
    file: '/tmp/gnexec-cov/t-toupper.ts',
    input: { s: 'hi' },
    expect: 'HI',
    lang: 'ts',
  },
  {
    label: 'cov-join',
    file: '/tmp/gnexec-cov/t-join.ts',
    input: { xs: ['a', 'b'] },
    expect: 'a,b',
    lang: 'ts',
  },
  {
    label: 'cov-reduce',
    file: '/tmp/gnexec-cov/t-reduce.ts',
    input: { xs: [1, 2, 3] },
    expect: '6',
    lang: 'ts',
  },
  {
    label: 'cov-map-named',
    file: '/tmp/gnexec-cov/t-map-named.ts',
    input: { xs: [1, 2, 3] },
    expect: '[2,4,6]',
    lang: 'ts',
  },
  {
    label: 'cov-spread',
    file: '/tmp/gnexec-cov/t-spread.ts',
    input: { a: 1 },
    expect: '{"a":1,"b":9}',
    lang: 'ts',
  },

  /* ─── Round 1 hardening (deferred fixtures, use `main`) ───────────── */
  {
    label: 'def-trycatch',
    file: '/tmp/gnexec-deferred/trycatch.ts',
    input: { kind: 'boom' },
    expect: 'caught:exploded',
    lang: 'ts',
    exportName: 'main',
  },
  {
    label: 'def-trycatch-ok',
    file: '/tmp/gnexec-deferred/trycatch.ts',
    input: { kind: 'fine' },
    expect: 'ok:fine',
    lang: 'ts',
    exportName: 'main',
  },
  {
    label: 'def-typeof',
    file: '/tmp/gnexec-deferred/probe_typeof.ts',
    input: { n: 5, s: 'hi' },
    expect: 'number|string|object|undefined',
    lang: 'ts',
    exportName: 'main',
  },
  {
    label: 'def-elseif',
    file: '/tmp/gnexec-deferred/elseif.ts',
    input: { n: 5 },
    expect: 'small',
    lang: 'ts',
    exportName: 'main',
  },
  {
    label: 'def-py-literals',
    file: '/tmp/gnexec-deferred/probe_pylit.py',
    input: {},
    expect: '2|v|10',
    lang: 'py',
    exportName: 'main',
  },
  {
    label: 'def-py-ternary',
    file: '/tmp/gnexec-deferred/probe_pyternary.py',
    input: { n: 5 },
    expect: 'small',
    lang: 'py',
    exportName: 'main',
  },
  {
    label: 'def-py-slice',
    file: '/tmp/gnexec-deferred/probe_pyslice.py',
    input: {},
    expect: '[20,30]|ell',
    lang: 'py',
    exportName: 'main',
  },
  {
    label: 'def-py-multi',
    file: '/tmp/gnexec-deferred/probe_pymulti.py',
    input: {},
    expect: '1|2',
    lang: 'py',
    exportName: 'main',
  },
  {
    label: 'def-py-dict-get',
    file: '/tmp/gnexec-deferred/probe_pydict.py',
    input: {},
    expect: 'v|missing',
    lang: 'py',
    exportName: 'main',
  },
  {
    label: 'def-py-unary',
    file: '/tmp/gnexec-deferred/probe_unary.py',
    input: { n: 3 },
    expect: '-3|3|false|-4',
    lang: 'py',
    exportName: 'main',
  },
  {
    label: 'def-uncaught-throw',
    file: '/tmp/gnexec-deferred/uncaught_throw.ts',
    input: { x: -1 },
    // gnexec exits 1 with "Uncaught: negative-input" on stderr.
    expect: /Uncaught: negative-input/,
    lang: 'ts',
    exportName: 'main',
    kind: 'expected-throw',
  },

  /* ─── Round 2 fixtures (R2 — may not all exist yet) ───────────────── */
  {
    label: 'r2-process',
    file: '/tmp/gnexec-r2/t-process.ts',
    input: {},
    expect: /^(darwin|linux|win32)$/,
    lang: 'ts',
  },
  {
    label: 'r2-date',
    file: '/tmp/gnexec-r2/t-date.ts',
    input: {},
    expect: /^\d{10,}$/,
    lang: 'ts',
  },
  {
    label: 'r2-url',
    file: '/tmp/gnexec-r2/t-url.ts',
    input: { url: 'https://example.com/foo' },
    expect: 'example.com',
    lang: 'ts',
  },
  {
    label: 'r2-regex',
    file: '/tmp/gnexec-r2/t-regex.ts',
    input: { s: 'hi there' },
    expect: 'true',
    lang: 'ts',
  },
  {
    label: 'r2-uint8',
    file: '/tmp/gnexec-r2/t-uint8.ts',
    input: {},
    expect: '3',
    lang: 'ts',
  },
  {
    label: 'r2-destruct',
    file: '/tmp/gnexec-r2/t-destruct.ts',
    input: { a: 3, b: 4 },
    expect: '7',
    lang: 'ts',
  },
  {
    label: 'r2-destruct-default',
    file: '/tmp/gnexec-r2/t-destruct-default.ts',
    input: { a: 3 },
    expect: '3',
    lang: 'ts',
  },

  /* ─── Multi-file imports (likely not wired through CLI yet) ───────── */
  {
    label: 'multi-file-import',
    file: '/tmp/multi-file/multi-a.ts',
    input: { x: 5 },
    expect: '10',
    lang: 'ts',
    // If `--resolve-imports` isn't piped through, gnexec returns
    // exit=6 "function 'helper' not found in topology". Mark as a
    // known-deferred regression marker; flip kind→pass when wired.
    kind: 'expected-unsupported',
    expectedExit: 6,
  },

  /* ─── Multi-language fixtures (Go/Rust/Java) ──────────────────────── */
  // Go and Java echo now PASS end-to-end through gnexec — the IR-driven
  // walker + Path B's per-language IR extractors + param_names emission
  // unlocked them on 2026-04-25. Rust still fails because `format!()` is
  // emitted as kind:"raw" IR and gnexec's byte-range fallback re-parses
  // the Rust source through tree-sitter-typescript — TS sees `!` as the
  // postfix non-null assertion and bails. Real fix: kind:"raw" should
  // emit IrNotSupported (exit 4) for non-TS source rather than attempt
  // a byte-range parse with the wrong grammar.
  {
    label: 'go-echo',
    file: '/tmp/multi-lang/echo.go',
    input: { name: 'go' },
    expect: 'hello:go',
    lang: 'go',
    exportName: 'App',
  },
  {
    label: 'rust-echo',
    file: '/tmp/multi-lang/echo.rs',
    input: { name: 'rust' },
    expect: 'hello:rust',
    lang: 'rust',
    exportName: 'app',
    // Flipped from expected-unsupported → pass (2026-04-26) once the Rust
    // extractor grew structured IR for `macro_invocation`: `format!("hello:{}",
    // n)` now lowers to a `template` IR with text + expr parts, evaluated
    // through gnexec's existing eval_template. See rust_lang.rs::rust_macro_to_ir.
  },
  {
    label: 'java-echo',
    file: '/tmp/multi-lang/Echo.java',
    input: { name: 'java' },
    expect: 'hello:java',
    lang: 'java',
    exportName: 'App',
  },

  /* ─── Cross-module imports: TypeScript (gnexec doesn't pass
        --resolve-imports yet, so these fail at runtime today with
        "function 'helper' not found in topology" / exit=6. Marked
        expected-unsupported so the harness exits 0; flips to `pass`
        the moment gnexec wires --resolve-imports through.) ─────────── */
  {
    label: 'multi-ts-helper-only',
    file: '/tmp/gnexec-multi-ts/entry.ts',
    input: { n: 5, name: 'gnode' },
    expect: 'hello:gnode:10',
    lang: 'ts',
    kind: 'expected-unsupported',
    expectedExit: 6,
  },
  {
    // Cycle-safety smoke: a.ts imports b, b.ts imports a. Today gnexec
    // doesn't traverse imports so it fails before it can loop — exit=7
    // (type coercion against the unresolved `b`-side identifier). Once
    // --resolve-imports flows through, this case verifies that the
    // resolver doesn't infinite-loop on the cycle (flip to `pass` with
    // the right `expect` then).
    label: 'multi-ts-cycle-safe',
    file: '/tmp/gnexec-multi-cycle/a.ts',
    input: { x: 3 },
    expect: '10',
    lang: 'ts',
    kind: 'expected-unsupported',
    expectedExit: 7,
  },

  /* ─── Cross-module imports: Python ─── */
  {
    label: 'multi-py-helper',
    file: '/tmp/gnexec-multi-py/mod_a.py',
    input: { n: 5 },
    expect: '10',
    lang: 'py',
    kind: 'expected-unsupported',
    expectedExit: 6,
  },

  /* ─── Extractor IR regression net (runs gnosis-polyglot, NOT gnexec).
        These lock in that the orchestration GG carries structured IR
        (expr=, bind=, callee=, args=) for Go/Rust/Java/TS/Python.
        This axis is independent of gnexec evaluator support. ───────── */
  {
    label: 'ir-go-app',
    file: '/tmp/multi-lang/echo.go',
    expect: /\bexpr=/,
    lang: 'go',
    kind: 'extractor-emits-ir',
    polyglotArgs: ['--mode', 'orchestration', '--format', 'gg'],
  },
  {
    label: 'ir-go-bind-callee',
    file: '/tmp/multi-lang/echo.go',
    // Go fixture's App returns greet(input["name"]) — there's no `bind=`
    // in the simple echo, but the call expr must carry callee+args.
    // Property order in the GG label is unordered, so this checks each
    // marker is present somewhere in the stdout.
    expect: (s) =>
      /callee='greet'/.test(s) && /args='\[/.test(s),
    lang: 'go',
    kind: 'extractor-emits-ir',
    polyglotArgs: ['--mode', 'orchestration', '--format', 'gg'],
  },
  {
    label: 'ir-rust-bind',
    file: '/tmp/multi-lang/echo.rs',
    // Rust fixture's `app` has `let n = input.get("name").unwrap();
    // greet(n)` — exercises both `bind=` and `callee=`. Properties are
    // unordered inside the GG label braces, so check each marker is
    // present anywhere in the stdout.
    expect: (s) => /bind='n'/.test(s) && /callee='greet'/.test(s),
    lang: 'rust',
    kind: 'extractor-emits-ir',
    polyglotArgs: ['--mode', 'orchestration', '--format', 'gg'],
  },
  {
    label: 'ir-java-bind',
    file: '/tmp/multi-lang/Echo.java',
    expect: (s) => /bind='n'/.test(s) && /callee='greet'/.test(s),
    lang: 'java',
    kind: 'extractor-emits-ir',
    polyglotArgs: ['--mode', 'orchestration', '--format', 'gg'],
  },
  {
    label: 'ir-ts-multi-resolve',
    file: '/tmp/gnexec-multi-ts/entry.ts',
    // --resolve-imports must follow ./helper and ./util/format and emit
    // module_origin markers for all three. Locks in the polyglot-side
    // import resolver independently of gnexec.
    expect: /module_origin: .*helper\.ts/s,
    lang: 'ts',
    kind: 'extractor-emits-ir',
    polyglotArgs: [
      '--mode',
      'orchestration',
      '--format',
      'gg',
      '--resolve-imports',
    ],
  },
  {
    label: 'ir-py-multi-resolve',
    file: '/tmp/gnexec-multi-py/mod_a.py',
    expect: /module_origin: .*mod_b\.py/s,
    lang: 'py',
    kind: 'extractor-emits-ir',
    polyglotArgs: [
      '--mode',
      'orchestration',
      '--format',
      'gg',
      '--resolve-imports',
    ],
  },

  /* ─── gnot fixture (queue for gnot IR alignment). Today gnexec doesn't
        dispatch to the gnot binary at all, so polyglot returns "no
        extractor for file" / exit=2. Flip to `pass` once gnot's
        emit-orchestration-gg is wired to gnexec. ─────────────────── */
  {
    label: 'gnot-is-lane',
    file: '/tmp/gnexec-gnot/test.gnot',
    input: {},
    expect: 'hello:gnot',
    lang: 'gnot',
    kind: 'expected-unsupported',
    expectedExit: 2,
  },
];

/* -------------------------------------------------------------------------- */
/* Runner                                                                     */
/* -------------------------------------------------------------------------- */

function median(xs) {
  if (xs.length === 0) return null;
  const sorted = [...xs].sort((a, b) => a - b);
  const mid = Math.floor(sorted.length / 2);
  return sorted.length % 2 === 0
    ? (sorted[mid - 1] + sorted[mid]) / 2
    : sorted[mid];
}

function fileMeta(p) {
  if (!existsSync(p)) return { exists: false };
  const s = statSync(p);
  return {
    exists: true,
    size: s.size,
    mtime: s.mtime.toISOString(),
  };
}

function checkExpected(stdout, expected) {
  const trimmed = stdout.replace(/\n+$/, '');
  if (typeof expected === 'string') {
    return trimmed === expected;
  }
  if (expected instanceof RegExp) {
    return expected.test(trimmed);
  }
  if (typeof expected === 'function') {
    return Boolean(expected(trimmed));
  }
  return false;
}

function runOnce(binary, args, env = process.env) {
  const t0 = performance.now();
  const result = spawnSync(binary, args, {
    encoding: 'utf8',
    timeout: TIMEOUT_MS,
    env,
  });
  const t1 = performance.now();
  return {
    wallMs: t1 - t0,
    exit: result.status,
    signal: result.signal,
    stdout: result.stdout ?? '',
    stderr: result.stderr ?? '',
    timedOut: Boolean(result.error && result.error.code === 'ETIMEDOUT'),
  };
}

function categorize(tc, run) {
  if (run.timedOut) return 'timeout';
  if (run.signal) return `signal-${run.signal}`;

  if (tc.kind === 'expected-throw') {
    if (run.exit !== 1) return `regression-expected-throw-exit-${run.exit}`;
    if (!checkExpected(run.stderr, tc.expect)) {
      return 'regression-expected-throw-msg';
    }
    return 'pass';
  }
  if (tc.kind === 'expected-unsupported') {
    if (run.exit !== (tc.expectedExit ?? 4)) {
      return `regression-unsupported-exit-${run.exit}`;
    }
    return 'pass';
  }
  if (tc.kind === 'extractor-emits-ir') {
    if (run.exit !== 0) return `regression-extractor-exit-${run.exit}`;
    if (!checkExpected(run.stdout, tc.expect)) return 'regression-ir-missing';
    return 'pass';
  }

  if (run.exit !== 0) return `fail-exit-${run.exit}`;
  if (!checkExpected(run.stdout, tc.expect)) return 'wrong-output';
  return 'pass';
}

function runCase(tc, opts = {}) {
  const { useJsBridge = false } = opts;

  if (!existsSync(tc.file)) {
    return {
      ...tc,
      status: 'file-missing',
      runs: [],
      medianWallMs: null,
      stdout: '',
      stderr: '',
      exit: null,
    };
  }

  let binary;
  let args;
  if (useJsBridge) {
    binary = process.execPath; // node
    // Use tsx loader so .ts source files load directly.
    args = [
      '--import',
      'tsx',
      INVOKE_APP,
      tc.file,
      JSON.stringify(tc.input),
      tc.exportName ?? 'app',
    ];
  } else if (tc.kind === 'extractor-emits-ir') {
    // IR regression net: invoke gnosis-polyglot directly with the
    // case's polyglotArgs, then grep stdout for IR markers.
    if (!existsSync(GNOSIS_POLYGLOT)) {
      return {
        ...tc,
        status: 'binary-missing',
        runs: [],
        medianWallMs: null,
        stdout: '',
        stderr: '',
        exit: null,
      };
    }
    binary = GNOSIS_POLYGLOT;
    args = [tc.file, ...(tc.polyglotArgs ?? [])];
  } else {
    if (!existsSync(GNEXEC)) {
      return {
        ...tc,
        status: 'binary-missing',
        runs: [],
        medianWallMs: null,
        stdout: '',
        stderr: '',
        exit: null,
      };
    }
    binary = GNEXEC;
    args = ['--input-json', JSON.stringify(tc.input ?? {})];
    if (tc.exportName) args.push('--export', tc.exportName);
    args.push(tc.file);
  }

  const env = { ...process.env };
  if (useJsBridge) env.GNODE_NO_GNEXEC = '1';

  const runs = [];
  for (let i = 0; i < ITERATIONS; i += 1) {
    runs.push(runOnce(binary, args, env));
  }
  // Median of measured runs (drop the first as warmup).
  const measured = runs.slice(1);
  const medianWallMs = median(measured.map((r) => r.wallMs));

  const last = runs[runs.length - 1];
  const status = useJsBridge
    ? // For JS bridge we just want a coarse pass/fail.
      last.exit === 0 && checkExpected(last.stdout, tc.expect)
      ? 'pass'
      : `fail-exit-${last.exit}`
    : categorize(tc, last);

  return {
    label: tc.label,
    file: tc.file,
    lang: tc.lang,
    kind: tc.kind ?? 'pass',
    exportName: tc.exportName ?? 'app',
    status,
    medianWallMs,
    walls: measured.map((r) => r.wallMs),
    exit: last.exit,
    stdout: last.stdout.slice(0, 400),
    stderr: last.stderr.slice(0, 400),
  };
}

/* -------------------------------------------------------------------------- */
/* Markdown summary                                                           */
/* -------------------------------------------------------------------------- */

function fmtMs(ms) {
  if (ms == null) return '—';
  if (ms < 10) return `${ms.toFixed(2)} ms`;
  if (ms < 1000) return `${ms.toFixed(1)} ms`;
  return `${(ms / 1000).toFixed(2)} s`;
}

function renderMarkdown(report) {
  const lines = [];
  lines.push(`# ${MONSTER_RUNTIME_NAME} cross-language benchmark`);
  lines.push('');
  lines.push(`- timestamp: \`${report.timestamp}\``);
  lines.push(`- harness git rev: \`${report.gitRev ?? 'null'}\``);
  lines.push(
    `- ${MONSTER_RUNTIME_NAME}: ${
      report.binaries.gnexec.exists
        ? `${report.binaries.gnexec.size} bytes, mtime ${report.binaries.gnexec.mtime}`
        : 'MISSING'
    }`
  );
  lines.push(
    `- gnosis-polyglot lib: ${
      report.binaries.polyglotLib.exists
        ? `${report.binaries.polyglotLib.size} bytes, mtime ${report.binaries.polyglotLib.mtime}`
        : 'MISSING'
    }`
  );
  lines.push(
    `- becky: ${
      report.binaries.becky.exists
        ? `${report.binaries.becky.size} bytes, mtime ${report.binaries.becky.mtime}`
        : 'MISSING'
    }`
  );
  lines.push('');
  lines.push(
    `## Totals: ${report.summary.pass}/${report.summary.knownTotal} known cases pass · ${report.summary.skipped} skipped (file/binary missing)`
  );
  lines.push('');

  // Per-language breakdown
  lines.push('## By language');
  lines.push('');
  lines.push(
    '| lang | pass | total | pass-rate | median wall | runtime median | budget |'
  );
  lines.push(
    '|------|------|-------|-----------|-------------|----------------|--------|'
  );
  for (const [lang, agg] of Object.entries(report.byLanguage)) {
    const rate =
      agg.total === 0 ? '—' : `${((agg.pass / agg.total) * 100).toFixed(0)}%`;
    const budget = report.budgets?.languageBudgetMs?.[lang];
    const budgetCell = budget == null ? '—' : `< ${budget} ms`;
    lines.push(
      `| ${lang} | ${agg.pass} | ${agg.total} | ${rate} | ${fmtMs(
        agg.medianWallMs
      )} | ${fmtMs(agg.runtimeMedianWallMs)} | ${budgetCell} |`
    );
  }
  lines.push('');

  // Per-case table
  lines.push('## Per-case results');
  lines.push('');
  lines.push('| label | lang | export | status | median | exit |');
  lines.push('|-------|------|--------|--------|--------|------|');
  for (const c of report.cases) {
    lines.push(
      `| ${c.label} | ${c.lang} | ${c.exportName} | ${c.status} | ${fmtMs(c.medianWallMs)} | ${c.exit ?? '—'} |`
    );
  }
  lines.push('');

  // Slowest 5
  const slowest = [...report.cases]
    .filter((c) => c.medianWallMs != null)
    .sort((a, b) => b.medianWallMs - a.medianWallMs)
    .slice(0, 5);
  lines.push('## Slowest 5');
  lines.push('');
  lines.push('| label | lang | median | status |');
  lines.push('|-------|------|--------|--------|');
  for (const c of slowest) {
    lines.push(
      `| ${c.label} | ${c.lang} | ${fmtMs(c.medianWallMs)} | ${c.status} |`
    );
  }
  lines.push('');

  // JS-bridge floor
  if (report.jsBridge && report.jsBridge.length > 0) {
    lines.push('## JS-bridge floor (GNODE_NO_GNEXEC=1, via tsx)');
    lines.push('');
    lines.push(`| label | ${MONSTER_RUNTIME_NAME} median | js-bridge median | speedup | bridge status |`);
    lines.push('|-------|---------------|------------------|---------|----------------|');
    for (const row of report.jsBridge) {
      const speedup =
        row.gnexecWallMs && row.bridgeWallMs
          ? `${(row.bridgeWallMs / row.gnexecWallMs).toFixed(1)}×`
          : '—';
      lines.push(
        `| ${row.label} | ${fmtMs(row.gnexecWallMs)} | ${fmtMs(
          row.bridgeWallMs
        )} | ${speedup} | ${row.bridgeStatus} |`
      );
    }
    lines.push('');
  }

  // Regressions
  if (report.summary.regressions.length > 0) {
    lines.push('## REGRESSIONS');
    lines.push('');
    for (const r of report.summary.regressions) {
      lines.push(`- **${r.label}** (${r.lang}): ${r.status} — exit=${r.exit}`);
      if (r.stderr) lines.push(`  - stderr: \`${r.stderr.split('\n')[0]}\``);
    }
    lines.push('');
  }

  // Wall-time budget violations
  const violations = report.budgets?.violations ?? [];
  if (violations.length > 0) {
    lines.push('## WALL-TIME BUDGET VIOLATIONS');
    lines.push('');
    for (const v of violations) {
      if (v.kind === 'language-budget') {
        lines.push(
          `- lang=\`${v.lang}\` median=${v.observedMs.toFixed(2)} ms > budget=${v.budgetMs} ms`
        );
      } else {
        lines.push(
          `- case=\`${v.label}\` (${v.lang}) median=${v.observedMs.toFixed(2)} ms > ceiling=${v.budgetMs} ms`
        );
      }
    }
    lines.push('');
  }

  return lines.join('\n');
}

/* -------------------------------------------------------------------------- */
/* Main                                                                       */
/* -------------------------------------------------------------------------- */

function getGitRev() {
  const r = spawnSync('git', ['rev-parse', 'HEAD'], {
    cwd: repoRoot,
    encoding: 'utf8',
  });
  if (r.status === 0) return r.stdout.trim();
  return null;
}

function main() {
  const startedAt = new Date().toISOString();

  const binaries = {
    gnexec: fileMeta(GNEXEC),
    polyglotLib: fileMeta(POLYGLOT_LIB),
    becky: fileMeta(BECKY),
  };

  process.stderr.write(`# ${MONSTER_RUNTIME_NAME} cross-language bench\n`);
  process.stderr.write(`# binary: ${GNEXEC}\n`);
  process.stderr.write(
    `#   exists=${binaries.gnexec.exists} size=${binaries.gnexec.size ?? '?'} mtime=${binaries.gnexec.mtime ?? '?'}\n`
  );

  const results = [];
  for (const tc of cases) {
    process.stderr.write(`  ${tc.label.padEnd(28)} `);
    const r = runCase(tc);
    results.push(r);
    process.stderr.write(`${r.status.padEnd(20)} ${fmtMs(r.medianWallMs)}\n`);
  }

  // JS-bridge floor — only canonical TS baselines flagged `floor`.
  const jsBridge = [];
  for (const tc of cases) {
    if (!tc.floor) continue;
    process.stderr.write(`  [bridge] ${tc.label.padEnd(20)} `);
    const r = runCase(tc, { useJsBridge: true });
    const gnexecRow = results.find((x) => x.label === tc.label);
    jsBridge.push({
      label: tc.label,
      gnexecWallMs: gnexecRow?.medianWallMs ?? null,
      bridgeWallMs: r.medianWallMs,
      bridgeStatus: r.status,
    });
    process.stderr.write(`${r.status.padEnd(20)} ${fmtMs(r.medianWallMs)}\n`);
  }

  // Aggregations.
  // `medianWallMs` is computed from gnexec-runtime cases only (kind=pass
  // or expected-throw or expected-unsupported). Extractor-emits-ir cases
  // run a different binary on a different code path — keep them out of
  // the per-language wall budget so we don't conflate the two axes.
  const byLanguage = {};
  for (const c of results) {
    const lang = c.lang;
    if (!byLanguage[lang]) {
      byLanguage[lang] = {
        pass: 0,
        total: 0,
        walls: [],
        runtimeWalls: [],
      };
    }
    if (c.status === 'file-missing' || c.status === 'binary-missing') continue;
    byLanguage[lang].total += 1;
    if (c.status === 'pass') byLanguage[lang].pass += 1;
    if (c.medianWallMs != null) byLanguage[lang].walls.push(c.medianWallMs);
    // Runtime-only walls: drives the per-language budget gate.
    if (
      c.medianWallMs != null &&
      c.kind !== 'extractor-emits-ir' &&
      c.status === 'pass'
    ) {
      byLanguage[lang].runtimeWalls.push(c.medianWallMs);
    }
  }
  for (const lang of Object.keys(byLanguage)) {
    byLanguage[lang].medianWallMs = median(byLanguage[lang].walls);
    byLanguage[lang].runtimeMedianWallMs = median(
      byLanguage[lang].runtimeWalls
    );
    delete byLanguage[lang].walls;
    delete byLanguage[lang].runtimeWalls;
  }

  const known = results.filter(
    (c) => c.status !== 'file-missing' && c.status !== 'binary-missing'
  );
  const passed = known.filter((c) => c.status === 'pass');
  const skipped = results.filter(
    (c) => c.status === 'file-missing' || c.status === 'binary-missing'
  );
  const regressions = known.filter((c) => c.status !== 'pass');

  const allWalls = passed
    .map((c) => c.medianWallMs)
    .filter((x) => x != null);
  const overallMedianWallMs = median(allWalls);

  // Per-language wall-time budget gate. Hard-fails when any language's
  // gnexec-runtime median crosses LANGUAGE_BUDGET_MS[lang].
  const budgetViolations = [];
  for (const [lang, agg] of Object.entries(byLanguage)) {
    const ceiling = LANGUAGE_BUDGET_MS[lang];
    if (ceiling == null) continue;
    const observed = agg.runtimeMedianWallMs;
    if (observed == null) continue; // No runtime cases (pure IR).
    if (observed > ceiling) {
      budgetViolations.push({
        kind: 'language-budget',
        lang,
        observedMs: observed,
        budgetMs: ceiling,
      });
    }
  }

  // Per-case slow-regression ceilings. Catches cases that blow past their
  // optimization target (e.g. def-trycatch post-#32 must stay under 50 ms).
  for (const c of results) {
    const ceiling = CASE_SLOW_CEILING_MS[c.label];
    if (ceiling == null) continue;
    if (c.medianWallMs == null) continue;
    if (c.status !== 'pass') continue; // Already counted as a regression.
    if (c.medianWallMs > ceiling) {
      budgetViolations.push({
        kind: 'case-ceiling',
        label: c.label,
        lang: c.lang,
        observedMs: c.medianWallMs,
        budgetMs: ceiling,
      });
    }
  }

  const report = {
    timestamp: startedAt,
    gitRev: getGitRev(),
    iterations: ITERATIONS,
    binaries,
    budgets: {
      languageBudgetMs: LANGUAGE_BUDGET_MS,
      caseSlowCeilingMs: CASE_SLOW_CEILING_MS,
      violations: budgetViolations,
    },
    summary: {
      total: results.length,
      knownTotal: known.length,
      pass: passed.length,
      skipped: skipped.length,
      overallMedianWallMs,
      regressions: regressions.map((r) => ({
        label: r.label,
        lang: r.lang,
        status: r.status,
        exit: r.exit,
        stderr: r.stderr,
      })),
    },
    byLanguage,
    cases: results,
    jsBridge,
  };

  writeFileSync(RESULTS_JSON, JSON.stringify(report, null, 2));
  process.stderr.write(`\n# wrote ${RESULTS_JSON}\n\n`);

  // Markdown to stdout.
  process.stdout.write(renderMarkdown(report));
  process.stdout.write('\n');

  if (regressions.length > 0) {
    process.stderr.write(
      `\n# REGRESSIONS (${regressions.length}): ${regressions.map((r) => r.label).join(', ')}\n`
    );
    process.exit(1);
  }
  if (budgetViolations.length > 0) {
    process.stderr.write(`\n# WALL-TIME BUDGET VIOLATIONS\n`);
    for (const v of budgetViolations) {
      if (v.kind === 'language-budget') {
        process.stderr.write(
          `#   lang=${v.lang} median=${v.observedMs.toFixed(2)}ms > budget=${v.budgetMs}ms\n`
        );
      } else {
        process.stderr.write(
          `#   case=${v.label} (${v.lang}) median=${v.observedMs.toFixed(2)}ms > ceiling=${v.budgetMs}ms\n`
        );
      }
    }
    process.exit(2);
  }
  process.exit(0);
}

main();
