// @a0n/gnosis/gnode/benchmarks/taylor.ts
// Taylor's Sequence membership — the Phyle Tripod test. `n` is a Taylor number
// when it is reachable from all three Fibonacci-Lucas axes at once:
//   1. a sum of two Fibonacci numbers   fib(a) + fib(b)
//   2. a sum of two Lucas numbers        lucas(a) + lucas(b)
//   3. a Fibonacci times a Lucas         fib(a) * lucas(b)
// The keystone is 199 = fib(10)+fib(12) = lucas(9)+lucas(10) = fib(1)*lucas(11).
// Mirrors Gnosis/TaylorsSequence.lean and distributed-inference bench-jar-shootoff.
// The hello/fib companion for the monster + vCPU benchmark trio.
function fibLucas(limit: number): { f: number[]; l: number[] } {
  const f = [0, 1];
  while (f[f.length - 1] <= limit) {
    f.push(f[f.length - 1] + f[f.length - 2]);
  }
  const l = [2, 1];
  while (l[l.length - 1] <= limit) {
    l.push(l[l.length - 1] + l[l.length - 2]);
  }
  return { f, l };
}

function isFibSum(n: number, f: number[]): boolean {
  for (let a = 2; a < f.length; a += 1) {
    if (f[a] > n) {
      break;
    }
    for (let b = a; b < f.length; b += 1) {
      const s = f[a] + f[b];
      if (s === n) {
        return true;
      }
      if (s > n) {
        break;
      }
    }
  }
  return false;
}

function isLucasSum(n: number, l: number[]): boolean {
  for (let a = 2; a < l.length; a += 1) {
    if (l[a] > n) {
      break;
    }
    for (let b = a; b < l.length; b += 1) {
      const s = l[a] + l[b];
      if (s === n) {
        return true;
      }
      if (s > n) {
        break;
      }
    }
  }
  return false;
}

function isFibLucasProduct(n: number, f: number[], l: number[]): boolean {
  for (let a = 1; a < f.length; a += 1) {
    if (f[a] > n) {
      break;
    }
    for (let b = 1; b < l.length; b += 1) {
      const p = f[a] * l[b];
      if (p === n) {
        return true;
      }
      if (p > n) {
        break;
      }
    }
  }
  return false;
}

function isTripod(n: number, f: number[], l: number[]): boolean {
  return n > 0 && isFibSum(n, f) && isLucasSum(n, l) && isFibLucasProduct(n, f, l);
}

/**
 * Handles the gnosis app workflow.
 */
export function app(input: { n: number }): string {
  const n = Math.max(0, Math.trunc(input.n));
  const tables = fibLucas(n + 2);
  return String(isTripod(n, tables.f, tables.l));
}
