// @a0n/gnosis/gnode/benchmarks/where-does-time-go.mjs
#!/usr/bin/env node
/* global console, process */
// Where does gnode time go?
//
// Runs a small matrix of (entrypoint path) × (subcommand) and breaks down
// wall time into Node startup, wrapper boot, bundle require, runCli, and
// the bridge's own compile/cache work. Use this to decide which subcommands
// would actually win from a Rust-native rewrite.
//
// Usage:
//   node gnode/benchmarks/where-does-time-go.mjs
//   node gnode/benchmarks/where-does-time-go.mjs --iterations=10 --warmup=2

import { spawn, spawnSync } from 'node:child_process';
import { performance } from 'node:perf_hooks';
import { existsSync, readFileSync, statSync } from 'node:fs';
import path from 'node:path';
import { createInterface } from 'node:readline';
import { fileURLToPath, pathToFileURL } from 'node:url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const gnosisDir = path.resolve(__dirname, '..', '..');
const repoRoot = path.resolve(gnosisDir, '..', '..');
const gnodeJs = path.resolve(gnosisDir, 'bin/gnode.js');
const bundlePath = path.resolve(gnosisDir, 'dist/gnode/bridge-driver.bundle.mjs');
const rustGnode = path.resolve(gnosisDir, 'gnode/target/release/gnode');
const becky = path.resolve(gnosisDir, 'polyglot/target/release/becky');
const monster = path.resolve(gnosisDir, 'polyglot/target/release/monster');
const gnexec = path.resolve(gnosisDir, 'polyglot/target/release/gnexec');
const monsterRuntime = existsSync(monster) ? monster : gnexec;
const echoTs = path.resolve(__dirname, 'echo.ts');
const fibTs = path.resolve(__dirname, 'fib.ts');
const fanoutTs = path.resolve(__dirname, 'fanout.ts');
const fanoutGg = path.resolve(__dirname, 'fanout.gg');
const echoGg = path.resolve(__dirname, 'echo.gg');
const fibGg = path.resolve(__dirname, 'fib.gg');
const bettiGg = path.resolve(gnosisDir, 'betti.gg');

function readFlag(name, fallback) {
  const raw = process.argv.slice(2).find((a) => a.startsWith(`--${name}=`));
  if (!raw) return fallback;
  const n = Number.parseInt(raw.slice(name.length + 3), 10);
  return Number.isFinite(n) && n > 0 ? n : fallback;
}

const iterations = readFlag('iterations', 5);
const warmup = readFlag('warmup', 2);
const fanoIterations = readFlag('fano-iterations', 100_000);
const residentWindow = readFlag('resident-window', 1024);
const residentOnly = process.argv.includes('--resident-only');
const skipFano = process.argv.includes('--skip-fano');
const percentCTheoremIds = [
  'Gnosis.RuntimeSpeedOfLightPercent.cache_hit_is_luminal',
  'Gnosis.RuntimeSpeedOfLightPercent.recompute_is_subluminal',
  'Gnosis.RuntimeSpeedOfLightPercent.less_compute_more_c',
  'AckermannIsLightSpeed.runtime_percent_c_rides_the_front',
];
const percentCEvidence = {
  residentRows: [],
  fanoCache: null,
};

// ---------- subjects ----------

// path: { name, build(subcmd, file) -> [argv0, ...args], available }
const paths = {
  'wrapper-bundle': {
    label: 'node bin/gnode.js (bundle)',
    available: existsSync(gnodeJs),
    build: (sub, file) => [
      process.execPath,
      [gnodeJs, sub, file, '--trace-timings'],
      {},
    ],
  },
  'wrapper-trust': {
    label: 'node bin/gnode.js (trusted prewarm)',
    available: existsSync(gnodeJs),
    build: (sub, file) => [
      process.execPath,
      [gnodeJs, sub, file, '--trace-timings'],
      { GNODE_TRUST_PREWARMED_BUNDLE: '1' },
    ],
  },
  'wrapper-tsx': {
    label: 'node bin/gnode.js (tsx forced)',
    available: existsSync(gnodeJs),
    build: (sub, file) => [
      process.execPath,
      [gnodeJs, sub, file, '--trace-timings'],
      { GNODE_FORCE_TSX: '1' },
    ],
  },
  'bundle-direct': {
    label: 'node dist/.../bundle.mjs (no wrapper)',
    available: existsSync(bundlePath),
    // Bundle exports runCli; we shim a tiny ESM driver inline via -e.
    build: (sub, file) => [
      process.execPath,
      [
        '--input-type=module',
        '-e',
        `const b = await import(${JSON.stringify(pathToFileURL(bundlePath).href)}); process.exit(await b.runCli(process.argv.slice(1)) ?? 0);`,
        '--',
        sub,
        file,
        '--trace-timings',
      ],
      {},
    ],
  },
  rust: {
    label: 'gnode (Rust release)',
    // Rust clap doesn't currently accept --trace-timings as a global flag,
    // so we measure wall only here. Internal timings are unchanged from
    // wrapper-tsx since both shell out to the same TS bridge via tsx.
    available: existsSync(rustGnode),
    build: (sub, file) => [rustGnode, [sub, file], {}],
  },
  becky: {
    label: 'becky (Rust, .gg only)',
    // Native Rust GG analysis floor. Different workload from the others
    // (operates on .gg topology, not .ts source — no TS compile step), but
    // shows the achievable wall time for any subcommand that doesn't need
    // a JS engine.
    available: existsSync(becky),
    build: (_sub, file) => [becky, [file, '--summary'], {}],
  },
  'gnode-analyze': {
    label: 'gnode analyze (Rust → becky/polyglot)',
    // End-to-end through Rust gnode binary, which routes Analyze by extension:
    //   .gg → becky
    //   .ts → gnosis-polyglot
    // Tests the actual user-facing path; will be slower than direct
    // becky/polyglot by the cost of one extra exec.
    available: existsSync(rustGnode) && existsSync(becky),
    build: (_sub, file) => [rustGnode, ['analyze', file, '--summary'], {}],
  },
  'gnode-analyze-ts': {
    label: 'gnode analyze .ts (Rust → polyglot)',
    // Specifically the TS-via-polyglot path. Compares apples-to-apples with
    // the wrapper-* and bundle-direct rows: same input file (.ts source),
    // different question (Betty analysis vs ts-bridge compile), but same
    // ballpark of work. Answers "how fast can a Rust-native TS analysis go?"
    available: existsSync(rustGnode),
    build: (_sub, file) => [rustGnode, ['analyze', file, '--summary'], {}],
  },
  monster: {
    label: 'monster (Rust executor, no Node)',
    // End-to-end TS execution with no JS engine: polyglot orchestration GG
    // → Rust topology walker → tiny expression evaluator → result.
    // The fixture input follows the exported app contract for each file.
    // See gnexec.rs for the supported structural dialect.
    available: existsSync(monsterRuntime) && existsSync(becky),
    build: (_sub, file) => [
      monsterRuntime,
      [file, '--input-json', monsterInputJson(file), '--export', 'app'],
      {},
    ],
  },
};

const subcommands = [
  { name: 'compile-echo', sub: 'compile', file: echoTs },
  { name: 'schedule-echo', sub: 'schedule', file: echoTs },
  { name: 'compile-fib', sub: 'compile', file: fibTs },
];

function monsterInputJson(file) {
  return file === fibTs ? '{"n":20}' : '{"name":"gnode"}';
}

// becky doesn't run on .ts; we route it to the closest .gg equivalent.
const beckyFileMap = new Map([
  [echoTs, echoGg],
  [fibTs, fibGg],
]);

// ---------- runner ----------

const TRACE_WRAPPER_RE = /\[gnode wrapper\]\s*(\{[\s\S]*?\n\})/;
const TRACE_TIMINGS_RE = /\[gnode timings\]\s*(\{[\s\S]*?\n\})/;

function runOnce(argv, env) {
  const t0 = performance.now();
  const res = spawnSync(argv[0], argv[1], {
    env: { ...process.env, ...env },
    encoding: 'utf8',
    stdio: ['ignore', 'pipe', 'pipe'],
  });
  const wallMs = performance.now() - t0;
  const stderr = res.stderr ?? '';
  const stdout = res.stdout ?? '';

  let wrapper = null;
  let timings = null;
  const wrapperMatch = stderr.match(TRACE_WRAPPER_RE);
  if (wrapperMatch) {
    try { wrapper = JSON.parse(wrapperMatch[1]); } catch {}
  }
  const timingsMatch = stderr.match(TRACE_TIMINGS_RE);
  if (timingsMatch) {
    try { timings = JSON.parse(timingsMatch[1]); } catch {}
  }

  return {
    wallMs,
    status: res.status,
    ok: res.status === 0,
    wrapper,
    timings,
    stdout,
    stderr,
  };
}

function median(values) {
  const sorted = [...values].sort((a, b) => a - b);
  const m = Math.floor(sorted.length / 2);
  return sorted.length % 2 ? sorted[m] : (sorted[m - 1] + sorted[m]) / 2;
}

function fmt(n) {
  if (n === undefined || n === null || Number.isNaN(n)) return '   -  ';
  return n.toFixed(1).padStart(6);
}

function fmtUs(ms) {
  if (narrowMissing(ms)) return '   -  ';
  return (ms * 1000).toFixed(1).padStart(8);
}

function narrowMissing(n) {
  return n === undefined || n === null || Number.isNaN(n);
}

function parseBenchLine(report, prefix) {
  const line = report
    .split('\n')
    .find((candidate) => candidate.startsWith(`${prefix} `));
  if (!line) return null;

  const fields = {};
  for (const part of line.split(/\s+/).slice(1)) {
    const [key, rawValue] = part.split('=');
    if (!key || rawValue === undefined) continue;
    const numeric = Number(rawValue);
    fields[key] = Number.isFinite(numeric) ? numeric : rawValue;
  }
  return fields;
}

function fmtRate(n) {
  if (n === undefined || n === null || Number.isNaN(n)) return '-';
  if (n >= 1_000_000) return `${(n / 1_000_000).toFixed(2)}M/s`;
  if (n >= 1_000) return `${(n / 1_000).toFixed(2)}K/s`;
  return `${n.toFixed(0)}/s`;
}

function cPercent(transmit, compute) {
  const denom = transmit + compute;
  if (denom <= 0) return 0;
  return Math.max(0, Math.min(100, (100 * transmit) / denom));
}

function cPercentFromHitRate(hits, total) {
  if (total <= 0) return 0;
  return (100 * hits) / total;
}

function fmtPercentC(value) {
  return `${value.toFixed(value >= 99.995 || value < 1 ? 3 : 1)}%c`;
}

function runFanoDiademBench() {
  const gnosisUringDir = path.resolve(repoRoot, 'open-source/x-gnosis/gnosis-uring');
  const binaryPath = path.resolve(
    gnosisUringDir,
    'target/release/fano-route-bench'
  );
  const reportPath = path.resolve(
    gnosisUringDir,
    'target/fano-route-bench.txt'
  );

  if (!existsSync(binaryPath)) {
    if (existsSync(reportPath)) {
      return readFileSync(reportPath, 'utf8');
    }
    throw new Error(
      'release binary not built; run `pnpm run a0 -- run gnosis-uring:bench:fano --skip-nx-cache` first'
    );
  }

  const result = spawnSync(binaryPath, [String(fanoIterations)], {
    cwd: gnosisUringDir,
    env: process.env,
    encoding: 'utf8',
    stdio: ['ignore', 'pipe', 'pipe'],
  });
  if (result.status !== 0) {
    const stderr = (result.stderr ?? '').trim();
    const stdout = (result.stdout ?? '').trim();
    throw new Error(
      `gnosis-uring FANO bench failed: ${stderr || stdout || result.status}`
    );
  }
  return readFileSync(reportPath, 'utf8');
}

const FLOW_HEADER_SIZE = 10;
const FLOW_FLAG_FORK = 0b0000_0001;
const FLOW_FLAG_RACE = 0b0000_0010;
const FLOW_FLAG_FOLD = 0b0000_0100;
const FLOW_FLAG_VENT = 0b0000_1000;

function encodeFlowFrame({ streamId, sequence, flags, payload }) {
  const payloadBuffer = Buffer.isBuffer(payload) ? payload : Buffer.from(payload);
  const header = Buffer.alloc(FLOW_HEADER_SIZE);
  header.writeUInt16BE(streamId, 0);
  header.writeUInt32BE(sequence, 2);
  header[6] = flags;
  header[7] = (payloadBuffer.length >> 16) & 0xff;
  header[8] = (payloadBuffer.length >> 8) & 0xff;
  header[9] = payloadBuffer.length & 0xff;
  return Buffer.concat([header, payloadBuffer]);
}

function createFlowFrameReader(stream) {
  let buffer = Buffer.alloc(0);
  const frames = [];
  const waiters = [];

  function pump() {
    while (buffer.length >= FLOW_HEADER_SIZE) {
      const payloadLength = (buffer[7] << 16) | (buffer[8] << 8) | buffer[9];
      const frameLength = FLOW_HEADER_SIZE + payloadLength;
      if (buffer.length < frameLength) break;
      const frame = buffer.subarray(0, frameLength);
      buffer = buffer.subarray(frameLength);
      const payload = frame.subarray(FLOW_HEADER_SIZE);
      const waiter = waiters.shift();
      if (waiter) waiter.resolve(payload);
      else frames.push(payload);
    }
  }

  stream.on('data', (chunk) => {
    buffer = Buffer.concat([buffer, chunk]);
    pump();
  });
  stream.once('end', () => {
    while (waiters.length > 0) waiters.shift().resolve(null);
  });

  return function readPayload() {
    if (frames.length > 0) return Promise.resolve(frames.shift());
    return new Promise((resolve) => waiters.push({ resolve }));
  };
}

function closeResident(child) {
  child.stdin.end();
  return new Promise((resolve) => {
    child.once('close', resolve);
    child.kill();
  });
}

async function measureResidentJsonl({ name, sourcePath }) {
  const child = spawn(
    monsterRuntime,
    [sourcePath, '--export', 'app', '--resident-jsonl'],
    {
      cwd: gnosisDir,
      env: { ...process.env, GNEXEC_FOIL_BACKEND: '1' },
      stdio: ['pipe', 'pipe', 'pipe'],
    }
  );
  let stderr = '';
  child.stderr.setEncoding('utf8');
  child.stderr.on('data', (chunk) => { stderr += chunk; });
  const lines = createInterface({ input: child.stdout });
  const iterator = lines[Symbol.asyncIterator]();
  const wall = [];
  const inner = [];

  for (let i = 0; i < warmup + iterations; i += 1) {
    const nextLine = iterator.next();
    const started = performance.now();
    child.stdin.write('{"userId":"ada"}\n');
    const { value, done } = await nextLine;
    const elapsed = performance.now() - started;
    if (done) throw new Error(`${name} closed stdout: ${stderr}`);
    const parsed = JSON.parse(value);
    if (!parsed.ok || parsed.result !== 'ada:user|ada:profile') {
      throw new Error(`${name} bad response: ${value} ${stderr}`);
    }
    if (i >= warmup) {
      wall.push(elapsed);
      inner.push(Number(parsed.elapsed_ns) / 1_000_000);
    }
  }

  await closeResident(child);
  return { name, wall: median(wall), inner: median(inner) };
}

async function measureResidentJsonlCannon({ name, sourcePath }) {
  const child = spawn(
    monsterRuntime,
    [sourcePath, '--export', 'app', '--resident-jsonl'],
    {
      cwd: gnosisDir,
      env: { ...process.env, GNEXEC_FOIL_BACKEND: '1' },
      stdio: ['pipe', 'pipe', 'pipe'],
    }
  );
  let stderr = '';
  child.stderr.setEncoding('utf8');
  child.stderr.on('data', (chunk) => { stderr += chunk; });
  const lines = createInterface({ input: child.stdout });
  const iterator = lines[Symbol.asyncIterator]();
  const samples = [];
  const batches = warmup + iterations;

  for (let batch = 0; batch < batches; batch += 1) {
    const started = performance.now();
    for (let i = 0; i < residentWindow; i += 1) {
      child.stdin.write('{"userId":"ada"}\n');
    }
    for (let i = 0; i < residentWindow; i += 1) {
      const { value, done } = await iterator.next();
      if (done) throw new Error(`${name} closed stdout: ${stderr}`);
      const parsed = JSON.parse(value);
      if (!parsed.ok || parsed.result !== 'ada:user|ada:profile') {
        throw new Error(`${name} bad response: ${value} ${stderr}`);
      }
    }
    if (batch >= warmup) {
      samples.push((performance.now() - started) / residentWindow);
    }
  }

  await closeResident(child);
  return { name, wall: median(samples), inner: null };
}

async function measureResidentFlow({ name, sourcePath }) {
  const child = spawn(
    monsterRuntime,
    [sourcePath, '--export', 'app', '--resident-flow-stdio'],
    {
      cwd: gnosisDir,
      env: { ...process.env, GNEXEC_FOIL_BACKEND: '1' },
      stdio: ['pipe', 'pipe', 'pipe'],
    }
  );
  let stderr = '';
  let sequence = 0;
  child.stderr.setEncoding('utf8');
  child.stderr.on('data', (chunk) => { stderr += chunk; });
  const readPayload = createFlowFrameReader(child.stdout);
  const wall = [];

  for (let i = 0; i < warmup + iterations; i += 1) {
    const frame = encodeFlowFrame({
      streamId: 1,
      sequence,
      flags: FLOW_FLAG_FORK | FLOW_FLAG_RACE | FLOW_FLAG_FOLD | FLOW_FLAG_VENT,
      payload: Buffer.from('ada'),
    });
    sequence += 1;
    const started = performance.now();
    child.stdin.write(frame);
    const payload = await readPayload();
    const elapsed = performance.now() - started;
    if (payload === null) throw new Error(`${name} closed stdout: ${stderr}`);
    if (payload.toString('utf8') !== 'ada:user|ada:profile') {
      throw new Error(`${name} bad response: ${payload.toString('utf8')} ${stderr}`);
    }
    if (i >= warmup) wall.push(elapsed);
  }

  await closeResident(child);
  return { name, wall: median(wall), inner: null };
}

async function measureResidentFlowCannon({ name, sourcePath }) {
  const child = spawn(
    monsterRuntime,
    [sourcePath, '--export', 'app', '--resident-flow-stdio'],
    {
      cwd: gnosisDir,
      env: { ...process.env, GNEXEC_FOIL_BACKEND: '1' },
      stdio: ['pipe', 'pipe', 'pipe'],
    }
  );
  let stderr = '';
  let sequence = 0;
  child.stderr.setEncoding('utf8');
  child.stderr.on('data', (chunk) => { stderr += chunk; });
  const readPayload = createFlowFrameReader(child.stdout);
  const samples = [];
  const batches = warmup + iterations;

  for (let batch = 0; batch < batches; batch += 1) {
    const started = performance.now();
    for (let i = 0; i < residentWindow; i += 1) {
      child.stdin.write(encodeFlowFrame({
        streamId: 1,
        sequence,
        flags: FLOW_FLAG_FORK | FLOW_FLAG_RACE | FLOW_FLAG_FOLD | FLOW_FLAG_VENT,
        payload: Buffer.from('ada'),
      }));
      sequence += 1;
    }
    for (let i = 0; i < residentWindow; i += 1) {
      const payload = await readPayload();
      if (payload === null) throw new Error(`${name} closed stdout: ${stderr}`);
      if (payload.toString('utf8') !== 'ada:user|ada:profile') {
        throw new Error(`${name} bad response: ${payload.toString('utf8')} ${stderr}`);
      }
    }
    if (batch >= warmup) {
      samples.push((performance.now() - started) / residentWindow);
    }
  }

  await closeResident(child);
  return { name, wall: median(samples), inner: null };
}

// ---------- main ----------

console.log(`gnode timing breakdown — ${iterations} iters (after ${warmup} warmup), median ms`);
console.log();

// Two floors: process-spawn cost depends on which engine we boot.
if (!residentOnly) {
  const nodeFloor = [];
  for (let i = 0; i < warmup + iterations; i += 1) {
    const r = runOnce([process.execPath, ['-e', '0']], {});
    if (i >= warmup) nodeFloor.push(r.wallMs);
  }
  const rustFloor = [];
  if (existsSync(becky)) {
    for (let i = 0; i < warmup + iterations; i += 1) {
      const r = runOnce([becky, ['--help']], {});
      if (i >= warmup) rustFloor.push(r.wallMs);
    }
  }
  console.log(`floor: node -e '0'   median=${fmt(median(nodeFloor))} ms  (applies to JS paths)`);
  if (rustFloor.length) {
    console.log(`floor: becky --help  median=${fmt(median(rustFloor))} ms  (applies to Rust paths — no JS engine)`);
  }
  console.log();

  const header = [
    'path',
    'subcmd',
    'wall',
    'wrap.total',
    'bundleReq',
    'runCli',
    'compile',
    'cache',
  ].map((h) => h.padStart(11)).join(' ');
  console.log(header);
  console.log('-'.repeat(header.length));

  const rows = [];
  for (const [pathKey, p] of Object.entries(paths)) {
    if (!p.available) {
      console.log(`(skip ${pathKey}: not built)`);
      continue;
    }
    for (const cmd of subcommands) {
      const samples = {
        wall: [], wrapTotal: [], bundleReq: [], runCli: [], compile: [], cacheStatus: [],
      };
      let firstFail = null;
      // becky and the .gg-routing form of gnode analyze need .gg inputs; the
      // .ts-via-polyglot row keeps the original .ts file (no remap).
      const inputFile = (pathKey === 'becky' || pathKey === 'gnode-analyze')
        ? (beckyFileMap.get(cmd.file) ?? cmd.file)
        : cmd.file;
      for (let i = 0; i < warmup + iterations; i += 1) {
        const [bin, args, env] = p.build(cmd.sub, inputFile);
        const r = runOnce([bin, args], env);
        if (!r.ok && !firstFail) {
          firstFail = r;
        }
        if (i < warmup) continue;
        samples.wall.push(r.wallMs);
        if (r.wrapper) {
          samples.wrapTotal.push(r.wrapper.totalMs);
          if (typeof r.wrapper.bundleRequireMs === 'number') {
            samples.bundleReq.push(r.wrapper.bundleRequireMs);
          }
          if (typeof r.wrapper.childSpawnMs === 'number') {
            // tsx path: childSpawnMs is the meaningful number; bundleReq=0
            samples.bundleReq.push(r.wrapper.childSpawnMs);
          }
        }
        if (r.timings) {
          if (typeof r.timings.totalMs === 'number') samples.runCli.push(r.timings.totalMs);
          if (r.timings.artifact?.compileMs !== undefined) samples.compile.push(r.timings.artifact.compileMs);
          if (r.timings.cacheStatus) samples.cacheStatus.push(r.timings.cacheStatus);
        }
      }

      if (firstFail) {
        const tail = (firstFail.stderr || '').split('\n').filter(Boolean).slice(-2).join(' | ');
        console.log(`${pathKey.padStart(11)} ${cmd.name.padStart(11)}   FAIL status=${firstFail.status}  ${tail}`);
        continue;
      }

      const row = {
        path: pathKey,
        subcmd: cmd.name,
        wall: median(samples.wall),
        wrapTotal: samples.wrapTotal.length ? median(samples.wrapTotal) : null,
        bundleReq: samples.bundleReq.length ? median(samples.bundleReq) : null,
        runCli: samples.runCli.length ? median(samples.runCli) : null,
        compile: samples.compile.length ? median(samples.compile) : null,
        cache: samples.cacheStatus[0] ?? '?',
      };
      rows.push(row);

      console.log(
        [
          pathKey,
          cmd.name,
          fmt(row.wall),
          fmt(row.wrapTotal),
          fmt(row.bundleReq),
          fmt(row.runCli),
          fmt(row.compile),
          (row.cache || '?').toString(),
        ].map((v, i) => i < 2 ? v.padStart(11) : v.padStart(11)).join(' ')
      );
    }
  }

  console.log();
  console.log('legend:');
  console.log('  wall       = wall-clock time of the whole process');
  console.log('  wrap.total = time inside bin/gnode.js main()  (= load bundle + runCli + book-keeping)');
  console.log('  bundleReq  = import(bundle.mjs) load time, OR childSpawn time on the tsx path');
  console.log('  runCli     = total inside bridge-driver.runCli (parse + cache + compile + write)');
  console.log('  compile    = actual TS→GG compile work (0 = cache hit)');
  console.log('  cache      = hit/miss for the first measured iteration');
  console.log();
  if (existsSync(bundlePath)) {
    console.log(`Bundle size: ${(statSync(bundlePath).size / 1024).toFixed(1)} KiB`);
  }
}

if (existsSync(monsterRuntime)) {
  console.log();
  console.log('resident fanout hot path — median microseconds');
  const residentHeader = ['path', 'wall', 'monster.inner', 'wire+harness']
    .map((h) => h.padStart(18))
    .join(' ');
  console.log(residentHeader);
  console.log('-'.repeat(residentHeader.length));

  const residentRows = [
    await measureResidentJsonl({
      name: 'monster-resident',
      sourcePath: fanoutTs,
    }),
    await measureResidentJsonl({
      name: 'monster-gg-resident',
      sourcePath: fanoutGg,
    }),
    await measureResidentFlow({
      name: 'monster-flow-resident',
      sourcePath: fanoutTs,
    }),
    await measureResidentFlow({
      name: 'monster-gg-flow-resident',
      sourcePath: fanoutGg,
    }),
    await measureResidentJsonlCannon({
      name: `monster-jsonl-cannon-${residentWindow}`,
      sourcePath: fanoutTs,
    }),
    await measureResidentFlowCannon({
      name: `monster-flow-cannon-${residentWindow}`,
      sourcePath: fanoutTs,
    }),
    await measureResidentFlowCannon({
      name: `monster-gg-flow-cannon-${residentWindow}`,
      sourcePath: fanoutGg,
    }),
  ];
  percentCEvidence.residentRows = residentRows;

  for (const row of residentRows) {
    const wireHarness = row.inner === null ? null : row.wall - row.inner;
    console.log(
      [
        row.name.padStart(18),
        fmtUs(row.wall),
        fmtUs(row.inner),
        fmtUs(wireHarness),
      ].join(' ')
    );
  }
  console.log();
  console.log('resident note: JSONL rows report Monster inner elapsed_ns; Flow rows currently report wall only.');
  console.log(`resident cannon rows write ${residentWindow} requests before folding responses; wall is per request.`);
}

if (!skipFano) {
  console.log();
  console.log(`gnosis-uring FANO/Diadem hot path — ${fanoIterations} iterations`);
  try {
    const report = runFanoDiademBench();
    const route = parseBenchLine(report, 'route');
    const cache = parseBenchLine(report, 'cache');
    const strideSelect = parseBenchLine(report, 'stride_select');
    const strideRace = parseBenchLine(report, 'stride_race');

    if (!route || !cache || !strideSelect || !strideRace) {
      throw new Error('missing expected FANO bench rows');
    }
    percentCEvidence.fanoCache = cache;

    console.log(
      [
        'route'.padStart(16),
        `wall=${fmt(route.elapsed_ms)} ms`,
        `throughput=${fmtRate(route.routes_per_second)}`,
        `xor=${route.xor}`,
        `same_phase=${route.same_phase}`,
        `validation=${route.validation_required}`,
      ].join(' ')
    );
    console.log(
      [
        'cache'.padStart(16),
        `uncached=${fmt(cache.uncached_ms)} ms`,
        `cached=${fmt(cache.cached_ms)} ms`,
        `cached_rps=${fmtRate(cache.cached_rps)}`,
        `hits=${cache.fano_hits}`,
        `misses=${cache.fano_misses}`,
      ].join(' ')
    );
    console.log(
      [
        'stride_select'.padStart(16),
        `wall=${fmt(strideSelect.elapsed_ms)} ms`,
        `throughput=${fmtRate(strideSelect.ops_per_second)}`,
      ].join(' ')
    );
    console.log(
      [
        'stride_race'.padStart(16),
        `wall=${fmt(strideRace.elapsed_ms)} ms`,
        `throughput=${fmtRate(strideRace.ops_per_second)}`,
      ].join(' ')
    );
    console.log(
      'fano note: route/cache include Monster/FOIL Diadem admission metadata; stride rows isolate the LeechTurboCannon selector and inspectable race.'
    );
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    console.log(`(skip gnosis-uring FANO/Diadem: ${message})`);
  }
}

console.log();
console.log('percent-c report — proof-carried cache warmth');
console.log('formula: cache-warmth %c = cache hits / total requests * 100');
console.log('delivery probe: transmit / (transmit + compute) * 100 when inner compute is reported');
for (const theoremId of percentCTheoremIds) {
  console.log(`theorem: ${theoremId}`);
}
console.log(
  [
    'business-default'.padStart(18),
    'cache_warmth=5.0%c',
    'source=docs/BUSINESS BINDER/COMPANY_CATALOG.md',
  ].join(' ')
);
console.log(
  [
    'business-warmed'.padStart(18),
    'cache_warmth=60.0%c',
    'source=entropy+RF+resident documented target',
  ].join(' ')
);
console.log(
  [
    'luminal-hit'.padStart(18),
    'cache_warmth=100.000%c',
    'source=single cache hit / photon',
  ].join(' ')
);

if (percentCEvidence.fanoCache) {
  const hits = Number(percentCEvidence.fanoCache.fano_hits ?? 0);
  const misses = Number(percentCEvidence.fanoCache.fano_misses ?? 0);
  const total = hits + misses;
  console.log(
    [
      'fano-cache'.padStart(18),
      `cache_warmth=${fmtPercentC(cPercentFromHitRate(hits, total))}`,
      `hits=${hits}`,
      `misses=${misses}`,
      `total=${total}`,
    ].join(' ')
  );
}

for (const row of percentCEvidence.residentRows) {
  if (row.inner === null) continue;
  const transmit = Math.max(0, row.wall - row.inner);
  console.log(
    [
      row.name.padStart(18),
      `delivery_share=${fmtPercentC(cPercent(transmit, row.inner))}`,
      `transmit_us=${(transmit * 1000).toFixed(1)}`,
      `compute_us=${(row.inner * 1000).toFixed(1)}`,
      'note=single-delivery-share-not-cache-warmth',
    ].join(' ')
  );
}
