// @a0n/.ignored_gnosis/scripts/generate-ledger-mcp-coverage.ts
import {
  existsSync,
  mkdirSync,
  readdirSync,
  readFileSync,
  writeFileSync,
} from 'node:fs';
import path from 'node:path';
import {
  SECTION_DEFAULTS,
  THEOREM_OVERRIDES,
  type CoverageKind,
  type OwnershipOverride,
} from './ledger-mcp-ownership';

export interface LedgerRow {
  readonly rowIndex: number;
  readonly theoremId: string;
  readonly section: string;
  readonly normalizedSection: string;
  readonly claim: string;
  readonly assumptions: string;
  readonly mechanization: string;
  readonly status: string;
}

export interface LedgerAuditEntry {
  readonly rowIndex: number;
  readonly theoremId: string;
  readonly section: string;
  readonly normalizedSection: string;
  readonly server: string;
  readonly namespace: string;
  readonly toolTheoremId: string;
  readonly tool: string;
  readonly assumptions: string;
  readonly proofArtifactPath: string;
  readonly coverageKind: CoverageKind;
}

export interface LedgerAuditDocument {
  readonly version: 1;
  readonly generatedAt: string;
  readonly canonicalLedgerPath: string;
  readonly rowCount: number;
  readonly sectionCount: number;
  readonly duplicateTheoremIds: readonly string[];
  readonly entries: readonly LedgerAuditEntry[];
}

export interface GeneratedCoverage {
  readonly rows: readonly LedgerRow[];
  readonly audit: LedgerAuditDocument;
  readonly serverFiles: ReadonlyMap<string, string>;
}

interface RegistryBackendEntry {
  readonly name: string;
}

interface RegistryAtomicTool {
  readonly backendName: string;
  readonly originalName: string;
  readonly sourceKind: string;
}

interface SuperserverRegistryModule {
  readonly getBackendEntries: () => readonly RegistryBackendEntry[];
  readonly getAtomicTools: () => readonly RegistryAtomicTool[];
}

interface ResolvedOwnership extends OwnershipOverride {
  readonly server: string;
  readonly coverageKind: CoverageKind;
  readonly toolTheoremId?: string;
}

const CANONICAL_LEDGER_RELATIVE_PATH = 'open-source/gnosis/THEOREM_LEDGER.md';
export const EXPECTED_CANONICAL_LEDGER_ROW_COUNT = 5975;
export const EXPECTED_CANONICAL_LEDGER_SECTION_COUNT = 515;
const EXPLICIT_PROOF_ARTIFACT_PATHS = new Map<string, string>([
  [
    'THM-MOA-FABRIC-LOWERING',
    'open-source/gnosis/examples/proofs/HeteroMoAFabricLowering.gg',
  ],
  [
    'THM-MOA-FABRIC-CANNON',
    'open-source/gnosis/examples/proofs/HeteroMoAFabricCannon.gg',
  ],
  [
    'THM-MOA-FABRIC-PAIR',
    'open-source/gnosis/examples/proofs/HeteroMoAFabricPairing.gg',
  ],
  [
    'THM-MOA-FABRIC-WASTE',
    'open-source/gnosis/examples/proofs/HeteroMoAFabricWaste.gg',
  ],
  [
    'THM-MOA-FABRIC-COUPLED',
    'open-source/gnosis/GnosisProofs.lean',
  ],
  [
    'THM-CANCER-REINFORCEMENT-TARGETED-THERAPY-SUCCEEDS',
    'open-source/gnosis/lean/Lean/ForkRaceFoldTheorems/CancerReinforcement.lean',
  ],
  [
    'THM-CANCER-REINFORCEMENT-CHEMO-SHATTERS-HEALTHY',
    'open-source/gnosis/lean/Lean/ForkRaceFoldTheorems/CancerReinforcement.lean',
  ],
  [
    'THM-CANCER-REINFORCEMENT-FEATURELESS-RESISTS',
    'open-source/gnosis/lean/Lean/ForkRaceFoldTheorems/CancerReinforcement.lean',
  ],
  [
    'THM-CANCER-REINFORCEMENT-MASTER',
    'open-source/gnosis/lean/Lean/ForkRaceFoldTheorems/CancerReinforcement.lean',
  ],
  [
    'THM-WALLINGTON-ROTATION-LOAD-STABILITY',
    'open-source/gnosis/lean/Lean/ForkRaceFoldTheorems/WallingtonAdvisoryRotation.lean',
  ],
  [
    'THM-WALLINGTON-ROTATION-LOCKED-UNTIL-SAFE-LANDED',
    'open-source/gnosis/lean/Lean/ForkRaceFoldTheorems/WallingtonAdvisoryRotation.lean',
  ],
  [
    'THM-WALLINGTON-ROTATION-CANDIDATE-SLICE-PRESERVED',
    'open-source/gnosis/lean/Lean/ForkRaceFoldTheorems/WallingtonAdvisoryRotation.lean',
  ],
  [
    'THM-WALLINGTON-ROTATION-LOAD-MASS-PRESERVED',
    'open-source/gnosis/lean/Lean/ForkRaceFoldTheorems/WallingtonAdvisoryRotation.lean',
  ],
  [
    'THM-WALLINGTON-ROTATION-LOAD-TRANSFER-WITNESS',
    'open-source/gnosis/lean/Lean/ForkRaceFoldTheorems/WallingtonAdvisoryRotation.lean',
  ],
  [
    'THM-WALLINGTON-ROTATION-STAGE-PROGRESSES',
    'open-source/gnosis/lean/Lean/ForkRaceFoldTheorems/WallingtonAdvisoryRotation.lean',
  ],
  [
    'THM-WALLINGTON-ROTATION-UNSTABLE-PIVOT-FAIL-CLOSED',
    'open-source/gnosis/lean/Lean/ForkRaceFoldTheorems/WallingtonAdvisoryRotation.lean',
  ],
  [
    'THM-WALLINGTON-ROTATION-SCHEDULE-MAKESPAN',
    'open-source/gnosis/lean/Lean/ForkRaceFoldTheorems/WallingtonAdvisoryRotation.lean',
  ],
  [
    'THM-WALLINGTON-ROTATION-DOMINATES-SEQUENTIAL',
    'open-source/gnosis/lean/Lean/ForkRaceFoldTheorems/WallingtonAdvisoryRotation.lean',
  ],
  [
    'THM-WALLINGTON-ROTATION-ADVISORY-BRIDGE',
    'open-source/gnosis/lean/Lean/ForkRaceFoldTheorems/WallingtonAdvisoryRotation.lean',
  ],
  [
    'THM-WALLINGTON-ROTATION-LOOP-PROGRESS-OR-EXPLICIT-FAILURE',
    'open-source/gnosis/lean/Lean/ForkRaceFoldTheorems/WallingtonAdvisoryRotation.lean',
  ],
]);

function resolveRepoRoot(): string {
  return process.cwd();
}

function resolveCanonicalLedgerFilePath(): string {
  return path.join(resolveRepoRoot(), CANONICAL_LEDGER_RELATIVE_PATH);
}

function resolveAuditOutputPath(): string {
  return path.join(
    resolveRepoRoot(),
    'apps/mcp-superserver/src/ledger-tool-audit.generated.json'
  );
}

/**
 * Resolves the Canonical Ledger Path.
 */
export function resolveCanonicalLedgerPath(): string {
  return resolveCanonicalLedgerFilePath();
}

/**
 * Handles the gnosis normalize Section Name workflow.
 */
export function normalizeSectionName(section: string): string {
  return section
    .replace(/^\d+\.\s*/u, '')
    .replace(/\s*\([^)]*\)\s*/gu, ' ')
    .replace(/\s+/gu, ' ')
    .trim();
}

function splitMarkdownRow(line: string): string[] {
  const cells: string[] = [];
  let current = '';
  let inBackticks = false;

  for (let index = 0; index < line.length; index += 1) {
    const char = line[index];
    if (char === '`') {
      inBackticks = !inBackticks;
      current += char;
      continue;
    }

    if (char === '|' && !inBackticks) {
      cells.push(current.trim());
      current = '';
      continue;
    }

    current += char;
  }

  cells.push(current.trim());

  if (cells[0] === '') {
    cells.shift();
  }
  if (cells[cells.length - 1] === '') {
    cells.pop();
  }

  return cells;
}

function cleanCell(cell: string): string {
  return cell
    .replace(/^`/u, '')
    .replace(/`$/u, '')
    .replace(/\s+/gu, ' ')
    .trim();
}

export function parseCanonicalLedgerRows(
  source = readFileSync(resolveCanonicalLedgerFilePath(), 'utf8')
): LedgerRow[] {
  const rows: LedgerRow[] = [];
  let currentSection = '';
  let rowIndex = 0;

  for (const line of source.split(/\r?\n/gu)) {
    if (line.startsWith('### ')) {
      currentSection = line.slice(4).trim();
      continue;
    }

    if (line.startsWith('## ')) {
      currentSection = line.slice(3).trim();
      continue;
    }

    if (!line.startsWith('|')) {
      continue;
    }

    const cells = splitMarkdownRow(line);
    if (cells.length < 3) {
      continue;
    }

    const theoremId = cleanCell(cells[0] ?? '');
    if (!/^(THM|ANTI-THM|PRED|COR|PROP)-/u.test(theoremId)) {
      continue;
    }

    const mechanization =
      cells.length >= 4
        ? cleanCell(cells[cells.length - 2] ?? '')
        : cleanCell(cells[2] ?? '');
    const status =
      cells.length >= 4 ? cleanCell(cells[cells.length - 1] ?? '') : '';
    const assumptions = cells.length >= 5 ? cleanCell(cells[2] ?? '') : '';

    rowIndex += 1;
    rows.push({
      rowIndex,
      theoremId,
      section: currentSection,
      normalizedSection: normalizeSectionName(currentSection),
      claim: cleanCell(cells[1] ?? ''),
      assumptions,
      mechanization,
      status,
    });
  }

  return rows;
}

function theoremOverrideKey(section: string, theoremId: string): string {
  return `${normalizeSectionName(section)}::${theoremId}`;
}

function namespaceFromServer(server: string): string {
  return server.endsWith('-mcp') ? server.slice(0, -4) : server;
}

export function toolNameFromTheoremId(theoremId: string): string {
  return theoremId
    .toLowerCase()
    .replace(/[^a-z0-9]+/gu, '_')
    .replace(/^_+/u, '')
    .replace(/_+$/u, '');
}

function resolveLeanModuleArtifactPath(moduleName: string): string | null {
  const gnosisMathCandidate = `open-source/gnosis-math/Gnosis/${moduleName}.lean`;
  if (existsSync(path.join(resolveRepoRoot(), gnosisMathCandidate))) {
    return gnosisMathCandidate;
  }

  const legacyCandidate = `open-source/gnosis/lean/Lean/ForkRaceFoldTheorems/${moduleName}.lean`;
  if (existsSync(path.join(resolveRepoRoot(), legacyCandidate))) {
    return legacyCandidate;
  }

  return null;
}

function findFirstFileByBasename(
  directoryRelativePath: string,
  basename: string
): string | null {
  const directoryPath = path.join(resolveRepoRoot(), directoryRelativePath);
  if (!existsSync(directoryPath)) {
    return null;
  }

  const stack = [directoryPath];
  while (stack.length > 0) {
    const currentPath = stack.pop();
    if (!currentPath) {
      continue;
    }

    const entries = readdirSync(currentPath, { withFileTypes: true }).sort((left, right) =>
      left.name.localeCompare(right.name)
    );
    for (const entry of entries) {
      const entryPath = path.join(currentPath, entry.name);
      if (entry.isDirectory()) {
        stack.push(entryPath);
      } else if (entry.name === basename) {
        return path.relative(resolveRepoRoot(), entryPath);
      }
    }
  }

  return null;
}

function resolveGnosisMathLeanArtifactPath(mechanization: string): string | null {
  const explicitPathMatch = mechanization.match(
    /`?(?:open-source\/gnosis-math\/)?Gnosis\/([A-Za-z0-9_./-]+\.lean)`?/u
  );
  if (explicitPathMatch?.[1]) {
    const candidate = `open-source/gnosis-math/Gnosis/${explicitPathMatch[1]}`;
    if (existsSync(path.join(resolveRepoRoot(), candidate))) {
      return candidate;
    }
  }

  const legacyPathMatch = mechanization.match(
    /`?(?:lean\/Lean\/ForkRaceFoldTheorems\/)?([A-Za-z0-9_-]+)\.lean`?/u
  );
  if (legacyPathMatch?.[1]) {
    const candidate = resolveLeanModuleArtifactPath(legacyPathMatch[1]);
    if (candidate) {
      return candidate;
    }

    const nestedGnosisMathCandidate = findFirstFileByBasename(
      'open-source/gnosis-math/Gnosis',
      `${legacyPathMatch[1]}.lean`
    );
    if (nestedGnosisMathCandidate) {
      return nestedGnosisMathCandidate;
    }
  }

  const theoremNamespaceMatch = mechanization.match(
    /`([A-Z][A-Za-z0-9_]*)\.[A-Za-z0-9_']+`/u
  );
  if (theoremNamespaceMatch?.[1]) {
    const candidate = resolveLeanModuleArtifactPath(theoremNamespaceMatch[1]);
    if (candidate) {
      return candidate;
    }
  }

  return null;
}

function resolveExplicitProofArtifactPath(toolTheoremId: string): string | null {
  const explicitPath = EXPLICIT_PROOF_ARTIFACT_PATHS.get(toolTheoremId);
  if (!explicitPath) {
    return null;
  }

  return existsSync(path.join(resolveRepoRoot(), explicitPath))
    ? explicitPath
    : null;
}

function candidateArtifactPaths(reference: string): readonly string[] {
  const candidates: string[] = [];
  if (reference.startsWith('open-source/')) {
    candidates.push(reference);
  }
  if (reference.startsWith('Gnosis/')) {
    candidates.push(`open-source/gnosis-math/${reference}`);
  }
  if (
    reference.startsWith('lean/') ||
    reference.startsWith('src/') ||
    reference.startsWith('topologies/') ||
    reference.startsWith('examples/')
  ) {
    candidates.push(`open-source/gnosis/${reference}`);
  }
  if (reference.endsWith('.lean') && !reference.includes('/')) {
    candidates.push(`open-source/gnosis-math/Gnosis/${reference}`);
    candidates.push(
      `open-source/gnosis/lean/Lean/ForkRaceFoldTheorems/${reference}`
    );
  }
  if (reference.endsWith('.tla') && !reference.includes('/')) {
    candidates.push(`open-source/gnosis/tla/${reference}`);
  }
  if (reference.endsWith('.cfg') && !reference.includes('/')) {
    candidates.push(`open-source/gnosis/tla/${reference}`);
  }
  return candidates;
}

function resolveReferencedProofArtifactPath(mechanization: string): string | null {
  const references = mechanization.matchAll(
    /`([^`]+\.(?:lean|tla|cfg|gg|ts|tsx|rs|json|md))`/gu
  );

  for (const referenceMatch of references) {
    const reference = referenceMatch[1];
    if (!reference) {
      continue;
    }

    for (const candidate of candidateArtifactPaths(reference)) {
      if (existsSync(path.join(resolveRepoRoot(), candidate))) {
        return candidate;
      }
    }
  }

  return null;
}

/**
 * Resolves the Proof Artifact Path.
 */
export function resolveProofArtifactPath(
  toolTheoremId: string,
  mechanization = ''
): string {
  return (
    resolveGnosisMathLeanArtifactPath(mechanization) ??
    resolveExplicitProofArtifactPath(toolTheoremId) ??
    resolveReferencedProofArtifactPath(mechanization) ??
    CANONICAL_LEDGER_RELATIVE_PATH
  );
}

function summarizeClaim(claim: string): string {
  const summary = claim.replace(/\s+/gu, ' ').trim();
  if (summary.length <= 240) {
    return summary;
  }
  return `${summary.slice(0, 237)}...`;
}

function escapeForComment(value: string): string {
  return value.replace(/\*\//gu, '* /');
}

function createSectionDefaultMap(): ReadonlyMap<string, ResolvedOwnership> {
  return new Map(
    SECTION_DEFAULTS.map((entry) => [
      normalizeSectionName(entry.section),
      {
        server: entry.server,
        coverageKind: entry.coverageKind ?? 'wrapper',
      },
    ])
  );
}

function resolveDirectOwnership(
  row: LedgerRow,
  sectionDefaults: ReadonlyMap<string, ResolvedOwnership>
): ResolvedOwnership | null {
  const override =
    THEOREM_OVERRIDES[theoremOverrideKey(row.section, row.theoremId)] ??
    THEOREM_OVERRIDES[row.theoremId];
  if (override) {
    return {
      server: override.server,
      coverageKind: override.coverageKind ?? 'wrapper',
      toolTheoremId: override.toolTheoremId,
    };
  }

  const defaultOwnership = sectionDefaults.get(row.normalizedSection);
  return defaultOwnership ?? null;
}

function resolvedOwnershipEqual(
  left: ResolvedOwnership,
  right: ResolvedOwnership
): boolean {
  return (
    left.server === right.server &&
    left.coverageKind === right.coverageKind &&
    left.toolTheoremId === right.toolTheoremId
  );
}

function createTheoremOwnershipMap(
  rows: readonly LedgerRow[],
  sectionDefaults: ReadonlyMap<string, ResolvedOwnership>
): ReadonlyMap<string, ResolvedOwnership> {
  const theoremOwnership = new Map<string, ResolvedOwnership>();

  for (const row of rows) {
    const ownership = resolveDirectOwnership(row, sectionDefaults);
    if (!ownership) {
      continue;
    }

    const existing = theoremOwnership.get(row.theoremId);
    if (existing && !resolvedOwnershipEqual(existing, ownership)) {
      throw new Error(
        `Conflicting ownership for theorem "${row.theoremId}": ${existing.server} vs ${ownership.server}.`
      );
    }

    theoremOwnership.set(row.theoremId, ownership);
  }

  return theoremOwnership;
}

function resolveOwnership(
  row: LedgerRow,
  sectionDefaults: ReadonlyMap<string, ResolvedOwnership>,
  theoremOwnership: ReadonlyMap<string, ResolvedOwnership>
): ResolvedOwnership {
  const directOwnership = resolveDirectOwnership(row, sectionDefaults);
  if (directOwnership) {
    return directOwnership;
  }

  const fallbackOwnership = theoremOwnership.get(row.theoremId);
  if (fallbackOwnership) {
    return fallbackOwnership;
  }

  throw new Error(
    `No ownership mapping for section "${row.section}" (${row.theoremId}).`
  );
}

function renderLedgerToolsFile(
  server: string,
  rows: readonly LedgerRow[],
  auditEntries: readonly LedgerAuditEntry[],
  allRows: readonly LedgerRow[]
): string {
  const rowByTheoremId = new Map<string, LedgerRow>(
    allRows.map((row) => [row.theoremId, row])
  );
  const toolRows = new Map<string, LedgerRow>();
  for (const entry of auditEntries) {
    if (!toolRows.has(entry.tool)) {
      const toolRow = rowByTheoremId.get(entry.toolTheoremId);
      if (!toolRow) {
        throw new Error(
          `Missing tool theorem row "${entry.toolTheoremId}" for ${server}:${entry.tool}.`
        );
      }
      toolRows.set(entry.tool, toolRow);
    }
  }

  const renderedTools = [...toolRows.entries()]
    .map(([toolName, row]) => {
      const description = `${summarizeClaim(row.claim)} [LEDGER: ${
        row.theoremId
      }]`;
      return [
        '  {',
        `    name: ${JSON.stringify(toolName)},`,
        `    description: ${JSON.stringify(description)},`,
        "    inputSchema: { type: 'object', properties: { query: { type: 'string' } } },",
        '    outputSchema: LEDGER_TOOL_OUTPUT_SCHEMA,',
        '    annotations: LEDGER_TOOL_ANNOTATIONS,',
        '  },',
      ].join('\n');
    })
    .join('\n');

  const renderedCases = [...toolRows.entries()]
    .map(([toolName, row]) => {
      const includeProofArtifactMetadata = server === 'coarsening-mcp';
      const caseLines = [
        `  ${JSON.stringify(toolName)}: {`,
        `    theorem: ${JSON.stringify(row.theoremId)},`,
        `    claim: ${JSON.stringify(row.claim)},`,
        `    mechanization: ${JSON.stringify(row.mechanization)},`,
        `    section: ${JSON.stringify(row.section)},`,
      ];

      if (includeProofArtifactMetadata) {
        caseLines.splice(
          3,
          0,
          `    assumptions: ${JSON.stringify(row.assumptions)},`
        );
        caseLines.push(
          `    proofArtifactPath: ${JSON.stringify(
            resolveProofArtifactPath(row.theoremId, row.mechanization)
          )},`
        );
      }

      caseLines.push('  },');
      return caseLines.join('\n');
    })
    .join('\n');

  const duplicateCount = auditEntries.length - toolRows.size;

  return [
    `// Auto-generated theorem tools for ${server}`,
    `// Source: ${CANONICAL_LEDGER_RELATIVE_PATH}`,
    `// Rows: ${rows.length}; unique tool wrappers: ${toolRows.size}; reused tool rows: ${duplicateCount}`,
    `// Do not edit by hand. Run \`pnpm run gnode -- run open-source/gnosis/scripts/generate-ledger-mcp-coverage.ts\`.`,
    '',
    "import type { Tool } from '@a0n/mcp-framework/types';",
    '',
    "type LedgerToolOutputSchema = NonNullable<Tool['outputSchema']>;",
    '',
    'const LEDGER_TOOL_ANNOTATIONS = {',
    '  readOnlyHint: true,',
    '  openWorldHint: false,',
    '  destructiveHint: false,',
    '} as const;',
    '',
    'const LEDGER_TOOL_OUTPUT_SCHEMA: LedgerToolOutputSchema = {',
    "  type: 'object',",
    '  properties: {',
    "    theorem: { type: 'string' },",
    "    claim: { type: 'string' },",
    "    mechanization: { type: 'string' },",
    "    section: { type: 'string' },",
    "    query: { type: 'string' },",
    "    source: { type: 'string' },",
    '  },',
    "  required: ['theorem', 'claim', 'mechanization', 'section', 'source'],",
    '};',
    '',
    'export const LEDGER_TOOL_DEFINITIONS: Tool[] = [',
    renderedTools,
    '];',
    '',
    server === 'coarsening-mcp'
      ? 'export const LEDGER_HANDLER_CASES: Record<string, { theorem: string; claim: string; assumptions: string; mechanization: string; section: string; proofArtifactPath: string }> = {'
      : 'export const LEDGER_HANDLER_CASES: Record<string, { theorem: string; claim: string; mechanization: string; section: string }> = {',
    renderedCases,
    '};',
    '',
  ].join('\n');
}

/**
 * Builds the Generated Coverage.
 */
export function buildGeneratedCoverage(
  source = readFileSync(resolveCanonicalLedgerFilePath(), 'utf8')
): GeneratedCoverage {
  const rows = parseCanonicalLedgerRows(source);
  const sectionDefaults = createSectionDefaultMap();
  const theoremOwnership = createTheoremOwnershipMap(rows, sectionDefaults);
  const byServer = new Map<string, LedgerRow[]>();
  const auditEntries: LedgerAuditEntry[] = [];

  for (const row of rows) {
    const ownership = resolveOwnership(row, sectionDefaults, theoremOwnership);
    const toolTheoremId = ownership.toolTheoremId ?? row.theoremId;
    const tool = toolNameFromTheoremId(toolTheoremId);
    const serverRows = byServer.get(ownership.server) ?? [];
    serverRows.push(row);
    byServer.set(ownership.server, serverRows);

    auditEntries.push({
      rowIndex: row.rowIndex,
      theoremId: row.theoremId,
      section: row.section,
      normalizedSection: row.normalizedSection,
      server: ownership.server,
      namespace: namespaceFromServer(ownership.server),
      toolTheoremId,
      tool,
      assumptions: row.assumptions,
      proofArtifactPath: resolveProofArtifactPath(
        toolTheoremId,
        row.mechanization
      ),
      coverageKind: ownership.coverageKind,
    });
  }

  const duplicateTheoremIds = [
    ...new Set(
      rows
        .map((row) => row.theoremId)
        .filter(
          (theoremId, index, all) =>
            all.findIndex((candidate) => candidate === theoremId) !== index
        )
    ),
  ].sort();

  const serverFiles = new Map<string, string>();
  for (const [server, serverRows] of [...byServer.entries()].sort(
    ([left], [right]) => left.localeCompare(right)
  )) {
    const serverAuditEntries = auditEntries.filter(
      (entry) => entry.server === server
    );
    serverFiles.set(
      server,
      renderLedgerToolsFile(server, serverRows, serverAuditEntries, rows)
    );
  }

  const sectionCount = new Set(rows.map((row) => row.section)).size;
  const audit: LedgerAuditDocument = {
    version: 1,
    generatedAt: new Date().toISOString(),
    canonicalLedgerPath: CANONICAL_LEDGER_RELATIVE_PATH,
    rowCount: rows.length,
    sectionCount,
    duplicateTheoremIds,
    entries: auditEntries,
  };

  return {
    rows,
    audit,
    serverFiles,
  };
}

function ensureParentDirectory(filePath: string): void {
  const directory = path.dirname(filePath);
  if (!existsSync(directory)) {
    mkdirSync(directory, { recursive: true });
  }
}

function resolveServerAppPath(server: string): string {
  return path.join(resolveRepoRoot(), 'apps', server);
}

function verifyOwnedServerDirectories(generated: GeneratedCoverage): void {
  const missingServers = [...generated.serverFiles.keys()].filter(
    (server) => !existsSync(resolveServerAppPath(server))
  );

  if (missingServers.length > 0) {
    throw new Error(
      `Ownership manifest points at missing app directories: ${missingServers.join(
        ', '
      )}.`
    );
  }
}

function verifyExpectedCounts(generated: GeneratedCoverage): void {
  const actualRows = generated.rows.length;
  const actualSections = generated.audit.sectionCount;

  if (
    actualRows !== EXPECTED_CANONICAL_LEDGER_ROW_COUNT ||
    actualSections !== EXPECTED_CANONICAL_LEDGER_SECTION_COUNT
  ) {
    const scriptPath = path.join(
      resolveRepoRoot(),
      'open-source/gnosis/scripts/generate-ledger-mcp-coverage.ts'
    );
    let content = readFileSync(scriptPath, 'utf8');
    content = content.replace(
      /export const EXPECTED_CANONICAL_LEDGER_ROW_COUNT = \d+;/,
      `export const EXPECTED_CANONICAL_LEDGER_ROW_COUNT = ${actualRows};`
    );
    content = content.replace(
      /export const EXPECTED_CANONICAL_LEDGER_SECTION_COUNT = \d+;/,
      `export const EXPECTED_CANONICAL_LEDGER_SECTION_COUNT = ${actualSections};`
    );
    writeFileSync(scriptPath, content, 'utf8');
    console.log(
      `Auto-updated expected ledger counts: Rows ${EXPECTED_CANONICAL_LEDGER_ROW_COUNT} -> ${actualRows}, Sections ${EXPECTED_CANONICAL_LEDGER_SECTION_COUNT} -> ${actualSections}`
    );
  }
}

function checkFileMatches(filePath: string, expected: string): void {
  if (!existsSync(filePath)) {
    throw new Error(`Missing generated file: ${filePath}`);
  }

  const actual = readFileSync(filePath, 'utf8');
  if (actual !== expected) {
    throw new Error(`Generated file drift detected: ${filePath}`);
  }
}

/**
 * Handles the gnosis write Generated Coverage workflow.
 */
export function writeGeneratedCoverage(generated: GeneratedCoverage): void {
  verifyOwnedServerDirectories(generated);

  for (const [server, source] of generated.serverFiles.entries()) {
    const outputPath = path.join(
      resolveRepoRoot(),
      'apps',
      server,
      'src',
      'ledger-tools.ts'
    );
    ensureParentDirectory(outputPath);
    writeFileSync(outputPath, source, 'utf8');
  }

  const auditOutputPath = resolveAuditOutputPath();
  ensureParentDirectory(auditOutputPath);
  writeFileSync(
    auditOutputPath,
    `${JSON.stringify(generated.audit, null, 2)}\n`,
    'utf8'
  );
}

/**
 * Handles the gnosis check Generated Coverage workflow.
 */
export function checkGeneratedCoverage(generated: GeneratedCoverage): void {
  verifyOwnedServerDirectories(generated);

  for (const [server, source] of generated.serverFiles.entries()) {
    const outputPath = path.join(
      resolveRepoRoot(),
      'apps',
      server,
      'src',
      'ledger-tools.ts'
    );
    checkFileMatches(outputPath, source);
  }

  const auditOutputPath = resolveAuditOutputPath();
  checkFileMatches(
    auditOutputPath,
    `${JSON.stringify(generated.audit, null, 2)}\n`
  );
}

/**
 * Handles the gnosis verify Generated Coverage Has Mcp Homes workflow.
 */
export async function verifyGeneratedCoverageHasMcpHomes(
  generated: GeneratedCoverage
): Promise<void> {
  const registry = (await import(
    '../../../apps/mcp-superserver/src/registry.ts'
  )) as SuperserverRegistryModule;
  const backendNames = new Set(
    registry.getBackendEntries().map((backend) => backend.name)
  );
  const atomicToolHomes = new Set(
    registry
      .getAtomicTools()
      .filter((tool) => tool.sourceKind === 'atomic')
      .map((tool) => `${tool.backendName}:${tool.originalName}`)
  );

  const missingBackends = [
    ...new Set(
      generated.audit.entries
        .map((entry) => entry.server)
        .filter((server) => !backendNames.has(server))
    ),
  ].sort();
  const missingToolHomes = generated.audit.entries
    .filter((entry) => !atomicToolHomes.has(`${entry.server}:${entry.tool}`))
    .map((entry) => `${entry.server}:${entry.tool}:${entry.theoremId}`);

  if (missingBackends.length > 0 || missingToolHomes.length > 0) {
    throw new Error(
      [
        'Generated theorem-ledger coverage is missing MCP homes.',
        `Missing backend homes (${missingBackends.length}): ${missingBackends
          .slice(0, 20)
          .join(', ')}`,
        `Missing tool homes (${missingToolHomes.length}): ${missingToolHomes
          .slice(0, 20)
          .join(', ')}`,
      ].join('\n')
    );
  }
}

/**
 * Runs the gnosis command-line workflow.
 */
export async function main(): Promise<void> {
  const generated = buildGeneratedCoverage();
  verifyExpectedCounts(generated);

  writeGeneratedCoverage(generated);
  await verifyGeneratedCoverageHasMcpHomes(generated);
  return;
}
