// @a0n/.ignored_gnosis/scripts/ledger-mcp-ownership.ts
export type CoverageKind = 'wrapper' | 'semantic' | 'umbrella';

export interface SectionDefault {
  readonly section: string;
  readonly server: string;
  readonly coverageKind?: CoverageKind;
}

export interface OwnershipOverride {
  readonly server: string;
  readonly coverageKind?: CoverageKind;
  readonly toolTheoremId?: string;
}

export const SECTION_DEFAULTS: readonly SectionDefault[] = [
  {
    section: 'Training Saturation Readiness',
    server: 'verification-mcp',
  },
  {
    section: 'Verified Reconstruction From Training Saturation',
    server: 'verification-mcp',
  },
  { section: 'Cosmic Architecture Closure', server: 'verification-mcp' },
  { section: 'Statistical Mechanics Accounting', server: 'thermodynamics-mcp' },
  { section: 'Sovereign Sieve Interpreter', server: 'verification-mcp' },
  { section: 'Gamut Braid', server: 'topology-mcp' },
  { section: 'Dark Fork Hypothesis (2026-04-15)', server: 'science-mcp' },
  {
    section: 'Dark Sector Atomized Encryption (2026-04-17)',
    server: 'quantum-mcp',
  },
  {
    section:
      'Mesh compaction (`Fin n` graphs, independent of DSAE cost and softmax convexity) (2026-04-18)',
    server: 'topology-mcp',
  },
  {
    section: 'Mesh attention, Buleyean, and reconstruction stack (2026-04-18)',
    server: 'topology-mcp',
  },
  { section: 'Markov Manpower Dynamics (2026-04-22)', server: 'sociology-mcp' },
  {
    section: 'Dating Marketplace Dynamics (2026-04-22)',
    server: 'relationships-mcp',
  },
  { section: 'Poe — *The Power of Words* (1845)', server: 'philosophy-mcp' },
  {
    section: 'Leibniz — *calculus ratiocinator* (bounded fragment)',
    server: 'philosophy-mcp',
  },
  { section: 'Laplace — *démon* (discrete determinism)', server: 'science-mcp' },
  {
    section: 'Laplace × Thoth — the demon on the self-talk clock',
    server: 'science-mcp',
  },
  {
    section: 'Laplace × Thoth — the demon faces the erasure ledger',
    server: 'science-mcp',
  },
  {
    section: 'Laplace × Thoth — the trace demon (total retrodiction)',
    server: 'science-mcp',
  },
  {
    section: 'Laplace × Thoth — selective undo (the commutator opens)',
    server: 'science-mcp',
  },
  {
    section: 'Laplace × Thoth — the erase/erase commutator (index shift)',
    server: 'science-mcp',
  },
  {
    section: 'Laplace × Thoth — deep selective undo (any depth of writes)',
    server: 'science-mcp',
  },
  {
    section: 'Laplace × Thoth — the reindexing step (erase under erase)',
    server: 'science-mcp',
  },
  {
    section:
      'Laplace × Thoth — bubble invariance (transpositions invisible to retrodiction)',
    server: 'science-mcp',
  },
  {
    section: 'Laplace × Thoth — the depth lift (deep swaps reduce to the top)',
    server: 'science-mcp',
  },
  {
    section: 'Laplace × Thoth — well-indexed stacks (the licensing invariant)',
    server: 'science-mcp',
  },
  {
    section:
      'Laplace × Thoth — the bubble fold (depth-general selective undo, assembled)',
    server: 'science-mcp',
  },
  { section: 'Borges — *La biblioteca de Babel*', server: 'philosophy-mcp' },
  {
    section: 'SocioTechnicalSandbox (coupled discrete toy)',
    server: 'sociology-mcp',
  },
  {
    section: 'PlanetaryHomologySandbox (finite graphs; narrative hooks only)',
    server: 'science-mcp',
  },
  {
    section: 'Arrow, Kernelel, Consciousness (§10.6)',
    server: 'consciousness-mcp',
  },
  { section: 'Narrative Spine Status', server: 'ergodicity-mcp' },
  { section: '75. The Kernel Gap (§10.6.17)', server: 'god-formula-mcp' },
  {
    section: '81K. Black Hole Void StructuralErrorgularity (§10.6)',
    server: 'gravity-mcp',
  },
  {
    section: '81AZ. Physics from the Kernel Formula (§10.6)',
    server: 'science-mcp',
  },
  {
    section: '85A. Base Ten as Fork/Race/Fold Consequence (§20.2)',
    server: 'mathematics-mcp',
  },
  {
    section: '85B. Convergence Parity: When to Stop Training (§20.3)',
    server: 'mathematics-mcp',
  },
  {
    section: '85C. Generalized Parity: Structural Invariants (§20.3)',
    server: 'mathematics-mcp',
  },
  {
    section: '85D. RL Bottlenecks Resolution in Negative Space (§20.4)',
    server: 'buleyean-mcp',
  },
  { section: '111. The Kernel Formula (§15.30)', server: 'god-formula-mcp' },
  {
    section: 'storms.watch Path Action and Kernel Weight Consolidation (§30)',
    server: 'storms-mcp',
  },
  {
    section: 'Manifold StructuralError Diagnostic (2026-03-29)',
    server: 'topology-mcp',
  },
  {
    section: 'Manifold Cross-Pollination (2026-03-29)',
    server: 'topology-mcp',
  },
  { section: '81AE. Astrolabe Vacuum Unification', server: 'relativity-mcp' },
  {
    section: 'Spacetime Causality (Relativistic Mechanics on the Formal Surface)',
    server: 'relativity-mcp',
  },
  {
    section: 'CanonicalLedger / InvariantProtocols Operational Guarantees (2026-04-26)',
    server: 'protocol-mcp',
  },
  {
    section: 'Juggling Orbit Promotion (2026-04-10)',
    server: 'verification-mcp',
  },
  {
    section: 'Certified Allocation Kernel Promotion (2026-04-10)',
    server: 'verification-mcp',
  },
  {
    section: 'Advisory Allocation Kernel Promotion (2026-04-10)',
    server: 'verification-mcp',
  },
  {
    section: 'Juggling Chaos Orchestration Promotion (2026-04-10)',
    server: 'verification-mcp',
  },
  {
    section: 'Progressive mesh ↔ granular segregation bridge (2026-04-17)',
    server: 'verification-mcp',
  },
  {
    section: 'Transformation Family Promotion (2026-04-10)',
    server: 'verification-mcp',
  },
  {
    section: 'Cancer Reinforcement Promotion (2026-04-10)',
    server: 'verification-mcp',
  },
  {
    section: 'Wallington Advisory Rotation Promotion (2026-04-10)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Perpetual Run 68) (2026-04-08)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Perpetual Run 67) (2026-04-08)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Perpetual Run 66) (2026-04-08)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Perpetual Run 65) (2026-04-08)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Perpetual Run 64) (2026-04-08)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Perpetual Run 63) (2026-04-08)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Perpetual Run 62) (2026-04-08)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Perpetual Run 61) (2026-04-08)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Perpetual Run 60) (2026-04-08)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Perpetual Run 58) (2026-04-08)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Perpetual Run 57) (2026-04-08)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Perpetual Run 56) (2026-04-08)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Perpetual Run 55) (2026-04-08)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Perpetual Run 54) (2026-04-08)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Perpetual Run 53) (2026-04-08)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Perpetual Run 51) (2026-04-08)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Perpetual Run 50) (2026-04-08)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Perpetual Run 49) (2026-04-08)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Perpetual Run 48) (2026-04-08)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Perpetual Run 47) (2026-04-08)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Perpetual Run 45) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Perpetual Run 44) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 40) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 39) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 38) (2026-04-08)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 37) (2026-04-08)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 36) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 35) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 34) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 33) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 32) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 31) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 30) (2026-04-08)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 29) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 28) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 27) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 26) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 25) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 24) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle 12: Buley Machine Run 5) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle 12: Buley Machine Run 4) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle 12: Buley Machine Run 3) (2026-04-07)',
    server: 'verification-mcp',
  },

  {
    section:
      'Iterative Cross-Pollination Pass (Cycle 12: Buley Machine Run 2) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle 12: Buley Machine Run 1) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 23) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 22) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section: 'Cycle 12: Cross-Pollination Pass (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 21) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 19) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 20) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 18) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 17) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 14) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 16) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 15) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 14) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle 10 Buley Perpetual Machine Run 12) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 13) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 12) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle 10 Moonshots & Bridges) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 11) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle 10 Buley Perpetual Machine Run 11) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 10) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 9) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 6) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section: 'Cycle 10: Buley Perpetual Machine (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 5) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section: 'Cycle 10: Cross-Pollination Pass (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section: 'Cycle 11: Cross-Pollination Pass (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section: 'Iterative Cross-Pollination Pass (Cycle 9 Iterative Pass 4)',
    server: 'verification-mcp',
  },
  {
    section: 'Iterative Cross-Pollination Pass (Cycle Douglasii Run 4)',
    server: 'verification-mcp',
  },
  {
    section:
      "Iterative Cross-Pollination Pass (Cycle 2026-04-06) (Buley's Run 3)",
    server: 'verification-mcp',
  },
  {
    section: 'Tectonic Tensor Queue Kernel Bridge (§19)',
    server: 'queueing-mcp',
  },
  {
    section: 'Iterative Cross-Pollination Pass (Cycle 3 Moonshots)',
    server: 'verification-mcp',
  },
  {
    section: 'Iterative Cross-Pollination Pass (Cycle Douglasii)',
    server: 'verification-mcp',
  },
  {
    section: 'Moonshot Stall Resolvers (Douglasii)',
    server: 'verification-mcp',
  },
  {
    section: 'Iterative Cross-Pollination Pass 2026-04-06 (Cycle 2)',
    server: 'verification-mcp',
  },
  {
    section: "Iterative Cross-Pollination Pass (2026-04-06) (Buley's Run 2)",
    server: 'verification-mcp',
  },
  {
    section: "Iterative Cross-Pollination Pass (2026-04-06) (Buley's Run)",
    server: 'verification-mcp',
  },
  {
    section: 'Iterative Cross-Pollination Pass (2026-04-06)',
    server: 'verification-mcp',
  },
  {
    section: 'Foundation: Fork/Race/Fold Axioms & Core Laws',
    server: 'verification-mcp',
  },
  {
    section: 'Monoidal Coherence & Traced Structure (Tracks Zeta + Eta)',
    server: 'coherence-mcp',
  },
  {
    section: 'Failure Theory: Collapse, Cost, and Control',
    server: 'failure-mcp',
  },
  {
    section: 'Thermodynamic Bridge: Erasure, Heat, and Information',
    server: 'thermodynamics-mcp',
  },
  { section: 'Beauty Optimality', server: 'beauty-mcp' },
  {
    section: 'Covering Space & Topological Deficit (Track Theta)',
    server: 'topology-mcp',
  },
  { section: 'Protocol, Settlement & Band Gap', server: 'protocol-mcp' },
  { section: 'Pipeline Scheduling (Section 7)', server: 'pipeline-mcp' },
  { section: 'Queueing Theory', server: 'queueing-mcp' },
  {
    section: 'Geometric & Compositional Ergodicity (Tracks Delta + Iota)',
    server: 'ergodicity-mcp',
  },
  {
    section: 'Lyapunov Synthesis (Tracks Kappa + Nu + Xi)',
    server: 'ergodicity-mcp',
  },
  { section: 'Envelope Convergence (Track Mu)', server: 'ergodicity-mcp' },
  {
    section: 'Wallington Rotation Optimality (Track Lambda)',
    server: 'pipeline-mcp',
  },
  { section: 'Race Winner Correctness (Track Omicron)', server: 'race-mcp' },
  { section: 'Frame Native Execution', server: 'protocol-mcp' },
  { section: 'Bitwise WASM Operations', server: 'bitwise-click' },
  { section: 'Void Walking', server: 'void-walking-mcp' },
  { section: 'Semiotic Deficit & Peace (Track Pi)', server: 'conflict-mcp' },
  { section: 'Negotiation Equilibrium', server: 'negotiation-mcp' },
  { section: 'Void Relativity & Grand Unification', server: 'relativity-mcp' },
  { section: 'Daisy Chain, Glossolalia & MOA', server: 'metacognition-mcp' },
  { section: 'Whip Wave Duality', server: 'protocol-mcp' },
  { section: 'Quorum Protocols', server: 'quorum-mcp' },
  {
    section: 'Cross-Language Semantic Type Theory',
    server: 'verification-mcp',
  },
  { section: 'Gnosis Compiler Proofs', server: 'verification-mcp' },
  { section: 'Rich-Club Topology', server: 'verification-mcp' },
  { section: 'Physarum Flux Routing', server: 'verification-mcp' },
  {
    section: 'Interference, Renormalization & Coarsening Synthesis',
    server: 'coarsening-mcp',
  },
  { section: 'Community Dominance & Skyrms Nadir', server: 'coarsening-mcp' },
  { section: 'Heterogeneous MoA Fabric', server: 'coarsening-mcp' },
  { section: 'Molecular Topology & Genomics', server: 'genomics-mcp' },
  { section: 'Cancer Topology (§3.17)', server: 'genomics-mcp' },
  { section: 'Cancer Predictions (§19.9)', server: 'genomics-mcp' },
  {
    section: 'Cancer Predictions Round 3 (§19.9, Predictions 11-15)',
    server: 'genomics-mcp',
  },
  {
    section: 'Cancer Predictions Round 4 (§19.9, Predictions 26-30)',
    server: 'genomics-mcp',
  },
  {
    section: 'Cancer Predictions Round 5 (§19.9, Predictions 31-35)',
    server: 'genomics-mcp',
  },
  {
    section: 'Cancer Treatment Strategies (§19.23, Predictions 76-80)',
    server: 'genomics-mcp',
  },
  {
    section: 'Cryptographic Predictions (§19.28, Predictions 101-105)',
    server: 'cybersecurity-mcp',
  },
  {
    section: 'Round 10 Promotion Closure (Round10PromotionClosure.lean)',
    server: 'negotiation-mcp',
  },
  { section: 'Quantum Cancer Triples (§19.9)', server: 'genomics-mcp' },
  { section: 'Cancer Confinement (§19.9 / §3.17)', server: 'genomics-mcp' },
  { section: 'Quantum Observer', server: 'quantum-mcp' },
  { section: 'Server Optimality & Codec Racing', server: 'race-mcp' },
  { section: 'Self-Hosting Optimality', server: 'void-walking-mcp' },
  { section: 'Humans Are Compilers', server: 'metacognition-mcp' },
  { section: 'Optimality Undecidable', server: 'race-mcp' },
  { section: 'The God Gap', server: 'race-mcp' },
  { section: 'Sliver-Exploration Theorem', server: 'coarsening-mcp' },
  { section: 'The Exploration Identity', server: 'coarsening-mcp' },
  { section: 'Buleyean Spin Pairing', server: 'buleyean-mcp' },
  { section: 'Particles Exist', server: 'quantum-mcp' },
  { section: 'Syzygy', server: 'topological-contagion-syzygy-mcp' },
  { section: 'Quark Confinement', server: 'quantum-mcp' },
  {
    section: 'The Ten Bosons: A Gnostic Particle Model',
    server: 'quantum-mcp',
  },
  { section: 'Ten-Mode Unification', server: 'quantum-mcp' },
  { section: 'The Gnostic Numbers', server: 'buleyean-mcp' },
  { section: 'Sleep Debt', server: 'sleep-mcp' },
  { section: 'Retrocausal Bound', server: 'quantum-mcp' },
  { section: 'American Frontier', server: 'topology-mcp' },
  { section: 'The Last Question (Asimov Bridge)', server: 'coherence-mcp' },
  { section: 'Chaitin Omega (Computability Bridge)', server: 'coherence-mcp' },
  {
    section: 'Non-Empirical Prediction (§15.22)',
    server: 'inference-forms-mcp',
  },
  { section: 'Grandfather Paradox (§15.23)', server: 'quantum-mcp' },
  {
    section: 'Five Novel AI Inference Forms (§15.24)',
    server: 'inference-forms-mcp',
  },
  {
    section: 'Novel Cross-Domain Predictions (§19.11)',
    server: 'inference-forms-mcp',
  },
  {
    section: 'Metacognitive Walker & Void Attention',
    server: 'metacognition-mcp',
  },
  { section: 'Reynolds BFT', server: 'quorum-mcp' },
  { section: 'Pluralist Republic', server: 'governance-mcp' },
  { section: 'Buleyean Evidence Standards (§28)', server: 'evidence-mcp' },
  {
    section:
      'Predictions Round 8: Memory, Ecology, Supply Chain, Jury, Skill Transfer (§19.27)',
    server: 'inference-forms-mcp',
  },
  { section: 'Arrow, Godel, Consciousness (§10.6)', server: 'coherence-mcp' },
  { section: 'Buleyean Pulse Quadrant (§10.6)', server: 'buleyean-mcp' },
  { section: 'Community Compositions (§10.6)', server: 'governance-mcp' },
  { section: 'CMB Visibility Boundary', server: 'science-mcp' },
  { section: 'Cosmic Optimal Delta', server: 'science-mcp' },
  { section: 'Knowability Split', server: 'science-mcp' },
  { section: 'Observable Universe Finitude', server: 'science-mcp' },
  {
    section: '81A. Celestial Projection Boundary (§10.6)',
    server: 'science-mcp',
  },
  {
    section: '81B. Celestial Knowability Boundary (§10.6)',
    server: 'science-mcp',
  },
  {
    section: '81C. Celestial Identifiability Boundary (§10.6)',
    server: 'science-mcp',
  },
  {
    section: '81D. Celestial Classifier Barrier (§10.6)',
    server: 'science-mcp',
  },
  {
    section: '81E. Celestial Orbit Prediction Boundary (§10.6)',
    server: 'science-mcp',
  },
  {
    section: '81F. Celestial Planet Taxonomy (§10.6)',
    server: 'science-mcp',
  },
  {
    section: '81G. Celestial Off-By-One Taxonomy and Universe Shape (§10.6)',
    server: 'science-mcp',
  },
  {
    section: '81H. Celestial Gain / Control Prediction (§10.6)',
    server: 'science-mcp',
  },
  { section: '81I. Celestial Survey Search (§10.6)', server: 'science-mcp' },
  {
    section: '81J. Celestial Natural Division (§10.6)',
    server: 'science-mcp',
  },
  {
    section: '81K. Black Hole Void Singularity (§10.6)',
    server: 'science-mcp',
  },
  { section: '81L. Fibonacci Size Surface (§10.6)', server: 'science-mcp' },
  {
    section: '81M. Photon Plus Sliver Frontier (§10.6)',
    server: 'science-mcp',
  },
  { section: '81N. Nested Wallington Tower (§10.6)', server: 'topology-mcp' },
  {
    section: '81O. Wallington Surface Admissibility (§10.6)',
    server: 'topology-mcp',
  },
  { section: '81P. Knowable Universe Map (§10.6)', server: 'science-mcp' },
  { section: '81Q. Atlas Coordinate Contract (§10.6)', server: 'topology-mcp' },
  {
    section: '81R. Stronger Cosmology Closure (§10.6)',
    server: 'relativity-mcp',
  },
  { section: '81S. Matter Explanation Closure (§10.6)', server: 'quantum-mcp' },
  { section: '81T. Proton Rest-Mass Candidate (§10.6)', server: 'quantum-mcp' },
  {
    section: '81U. Proton Calibration Boundary (§10.6)',
    server: 'quantum-mcp',
  },
  {
    section: '81V. Proton Quantized Uniqueness (§10.6)',
    server: 'quantum-mcp',
  },
  {
    section: '81W. Particle Theory Core Closure (§10.6)',
    server: 'quantum-mcp',
  },
  { section: '81X. Fermion Charge Closure (§10.6)', server: 'quantum-mcp' },
  { section: '81Y. Flavor Mixing Closure (§10.6)', server: 'quantum-mcp' },
  { section: '81Z. Running Coupling Closure (§10.6)', server: 'quantum-mcp' },
  {
    section: '81AA. Electroweak Scattering Closure (§10.6)',
    server: 'quantum-mcp',
  },
  {
    section: '81AB. Standard Model Shell Closure (§10.6)',
    server: 'science-mcp',
  },
  { section: '81AC. Flavor Phase Closure (§10.6)', server: 'quantum-mcp' },
  {
    section: '81AD. Anomaly Cancellation Closure (§10.6)',
    server: 'quantum-mcp',
  },
  {
    section: '81AE. Continuous Beta Function Closure (§10.6)',
    server: 'quantum-mcp',
  },
  {
    section: '81AF. Scattering Amplitude Closure (§10.6)',
    server: 'quantum-mcp',
  },
  {
    section: '81AG. Extended Standard Model Shell Closure (§10.6)',
    server: 'science-mcp',
  },
  {
    section: '81AH. Gauge Representation Closure (§10.6)',
    server: 'quantum-mcp',
  },
  {
    section: '81AI. Yukawa Mass Generation Closure (§10.6)',
    server: 'quantum-mcp',
  },
  { section: '81AJ. Gauge Mass Shell Closure (§10.6)', server: 'quantum-mcp' },
  { section: '81AK. Measured Flavor Closure (§10.6)', server: 'quantum-mcp' },
  { section: '81AL. Perturbative Beta Closure (§10.6)', server: 'quantum-mcp' },
  {
    section: '81AM. Kinematic Scattering Closure (§10.6)',
    server: 'quantum-mcp',
  },
  {
    section: '81AN. Gravitational Dynamics Closure (§10.6)',
    server: 'relativity-mcp',
  },
  {
    section: '81AO. Residual Strong Force Closure (§10.6)',
    server: 'quantum-mcp',
  },
  {
    section: '81AP. Unified Replacement Closure (§10.6)',
    server: 'science-mcp',
  },
  {
    section: '81AQ. Standard Model Replacement Sweep (§21)',
    server: 'science-mcp',
  },
  {
    section: '81AR. P ≠ NP Structural Separation (§10.6)',
    server: 'verification-mcp',
  },
  { section: '81AS. Hierarchy Problem Closure (§10.6)', server: 'science-mcp' },
  { section: '81AT. Strong CP Closure (§10.6)', server: 'quantum-mcp' },
  { section: '81AU. Yang-Mills Mass Gap (§10.6)', server: 'quantum-mcp' },
  { section: '81AV. Baryogenesis Closure (§10.6)', server: 'quantum-mcp' },
  {
    section: '81AW. Cosmological Constant Closure (§10.6)',
    server: 'relativity-mcp',
  },
  {
    section: '81AX. Dark Matter Identity Closure (§10.6)',
    server: 'science-mcp',
  },
  { section: '81AY. Dark Sector Closure (§10.6)', server: 'science-mcp' },
  {
    section: '81AZ. Physics from the God Formula (§10.6)',
    server: 'god-formula-mcp',
  },
  {
    section: '81BA. Physics Grand Unification (§10.6)',
    server: 'god-formula-mcp',
  },
  { section: '81BB. Precision Flavor Closure (§10.6)', server: 'quantum-mcp' },
  { section: 'Covering Space Causality (§3.3)', server: 'topology-mcp' },
  {
    section: 'Hodge Bridge & Algebraic Connectivity (Track Omega)',
    server: 'topology-mcp',
  },
  {
    section: 'Kakeya-Besicovitch Rotation (Track Omega)',
    server: 'topology-mcp',
  },
  {
    section: 'Moving Sofa Topology (Track Omega — Pass 67)',
    server: 'topology-mcp',
  },
  {
    section:
      'Glossolalia Mitigation — The Sofa Guard (Track Omega — Pass 67 Extension)',
    server: 'topology-mcp',
  },
  {
    section: 'Impossible Tilings and Aperiodicity (Track Omega — Pass 68)',
    server: 'topology-mcp',
  },
  {
    section:
      'Higher-Dimensional Covering Spaces x Aperiodic Projections (Track Omega — Pass 69)',
    server: 'topology-mcp',
  },
  { section: 'Cross-File Compositions (§19)', server: 'inference-forms-mcp' },
  { section: 'Cross-Module Identities (§19)', server: 'inference-forms-mcp' },
  { section: 'Deep Compositions (§19)', server: 'inference-forms-mcp' },
  { section: 'Diversity Unwound (§10.6)', server: 'topology-mcp' },
  { section: 'Evolution (§10.6)', server: 'genomics-mcp' },
  { section: 'Failure Controller (§9)', server: 'failure-mcp' },
  { section: 'Failure Pareto (§9)', server: 'failure-mcp' },
  { section: 'Failure Universality (§9)', server: 'failure-mcp' },
  { section: 'Fisher Manifold (§15)', server: 'buleyean-mcp' },
  { section: 'Grand Unification (§10.6)', server: 'relativity-mcp' },
  { section: 'Launch Offset Dominance (§12)', server: 'coarsening-mcp' },
  { section: 'Novel Compositions (§19)', server: 'inference-forms-mcp' },
  {
    section: 'Novel Predictions Round 28 (§19)',
    server: 'inference-forms-mcp',
  },
  { section: 'Post-Linear Transition (§10.6)', server: 'topology-mcp' },
  { section: 'Prediction Proofs Round 1 (§19)', server: 'inference-forms-mcp' },
  { section: 'Predictions Round 2 (§19)', server: 'inference-forms-mcp' },
  { section: 'Predictions Round 3 (§19)', server: 'inference-forms-mcp' },
  { section: 'Predictions Round 4 (§19)', server: 'inference-forms-mcp' },
  { section: 'Predictions Round 10 (§19)', server: 'inference-forms-mcp' },
  { section: 'Predictions Round 11 (§19)', server: 'inference-forms-mcp' },
  { section: 'Predictions Round 14 (§19)', server: 'inference-forms-mcp' },
  {
    section: 'Predictions Round 5 (§19, Predictions 66-70)',
    server: 'inference-forms-mcp',
  },
  {
    section: 'Predictions Round 6 (§19, Predictions 71-75)',
    server: 'inference-forms-mcp',
  },
  {
    section: 'Predictions Round 7 (§19, Predictions 111-115)',
    server: 'inference-forms-mcp',
  },
  {
    section: 'Predictions Round 9 (§19, Predictions 142-146)',
    server: 'inference-forms-mcp',
  },
  {
    section: 'Predictions Round 12 (§19, Predictions 177-181)',
    server: 'inference-forms-mcp',
  },
  { section: 'Predictions Round 13 (§19)', server: 'inference-forms-mcp' },
  {
    section: 'Adaptive Bisimulation Predictions (§19, Predictions 232-236)',
    server: 'inference-forms-mcp',
  },
  {
    section: 'Hetero Compositional Predictions (§19, Predictions 242-246)',
    server: 'inference-forms-mcp',
  },
  { section: 'Complement Oscillation (§15)', server: 'buleyean-mcp' },
  {
    section: 'Novel Triple Compositions (§19, Predictions 292-296)',
    server: 'inference-forms-mcp',
  },
  {
    section: 'Novel Triple Compositions Round 2 (§19, Predictions 297-301)',
    server: 'inference-forms-mcp',
  },
  {
    section: 'Teleportation Evidence Predictions (§19, Predictions 212-216)',
    server: 'inference-forms-mcp',
  },
  { section: 'Triple Compositions (§19)', server: 'inference-forms-mcp' },
  {
    section: 'Predictions Round 15: Sandwich-Derived Statistics (§19)',
    server: 'inference-forms-mcp',
  },
  {
    section: 'Predictions Round 16: Cross-Sandwich Combinatorics (§19)',
    server: 'inference-forms-mcp',
  },
  { section: 'Semiotic Triples (§19)', server: 'conflict-mcp' },
  { section: 'Sliver from Vent (§10.6)', server: 'thermodynamics-mcp' },
  { section: 'Sliver of Hope (§10.6)', server: 'buleyean-mcp' },
  { section: 'Statistical Teleportation (§15)', server: 'quantum-mcp' },
  {
    section: 'Deceptacon / DualVoidDeceptacon Bridge (§15.30.8)',
    server: 'metacognition-mcp',
  },
  {
    section: 'Deceptacon Head Elimination (§15.30.8)',
    server: 'metacognition-mcp',
  },
  { section: 'Time Travel Topology (§15.23)', server: 'quantum-mcp' },
  { section: 'Trade Topology (§19)', server: 'economics-mcp' },
  {
    section: 'Trade Topology Round 3 (§19, P212-P216)',
    server: 'economics-mcp',
  },
  {
    section: 'Trade Topology Round 4 (§19, P222-P226)',
    server: 'economics-mcp',
  },
  { section: 'Novel Inference Forms (§19.21)', server: 'inference-forms-mcp' },
  {
    section: 'Floor Theorems: Missing Lower Bounds (§3, §7, §9, §12, §13, §15)',
    server: 'verification-mcp',
  },
  {
    section:
      'Five-Parameter Void Walker Personality Model (ch17 section 15.10.7, 15.12)',
    server: 'metacognition-mcp',
  },
  {
    section: 'Buleyean Proof Topology System (ch17 section 20.2.8)',
    server: 'verification-mcp',
  },
  {
    section:
      'Buleyean Grand Unification: RL + Logic + Probability = Rejection (ch17 sections 15.10.7, 20.2.8)',
    server: 'buleyean-mcp',
  },
  // ── Combinatorial Brute Force (Sections 86-90) ──────────────────────
  {
    section:
      'Combinatorial Brute Force: Cross-Module Compositions (§17+§22+§23)',
    server: 'inference-forms-mcp',
  },
  {
    section:
      'Combinatorial Brute Force Round 2: Cross-Domain Compositions (§17+§19+§22+§23)',
    server: 'inference-forms-mcp',
  },
  {
    section:
      'Combinatorial Brute Force Round 3: Exotic Cross-Domain (§9+§17+§22+§23)',
    server: 'inference-forms-mcp',
  },
  {
    section:
      'Combinatorial Brute Force Round 4: Governance, Physics, and Waste (§7+§9+§11+§25+§26)',
    server: 'inference-forms-mcp',
  },
  {
    section:
      'Combinatorial Brute Force Round 5: Deep Quad Compositions and Grand Unification (§7+§9+§11+§17+§22+§23+§25+§26)',
    server: 'inference-forms-mcp',
  },
  // ── Philosophical Allegories & Logic (Sections 91-98) ────────────────
  {
    section:
      'Philosophical Allegories: Machine-Checked Ancient Wisdom (§20+§22+§23+§25)',
    server: 'philosophy-mcp',
  },
  { section: "I. Plato's Cave (Republic, Book VII)", server: 'philosophy-mcp' },
  { section: "II. Aristotle's Hamartia (Poetics)", server: 'philosophy-mcp' },
  {
    section: 'III. The Socratic Method (Gorgias, Meno)',
    server: 'philosophy-mcp',
  },
  {
    section: `III½. Cicero's Maxim ("Errare malo cum Platone")`,
    server: 'philosophy-mcp',
  },
  {
    section: "IV. Plato's Divided Line (Republic, Book VI)",
    server: 'philosophy-mcp',
  },
  {
    section: 'V. Buddhist Two Truths (Nagarjuna, Madhyamaka)',
    server: 'philosophy-mcp',
  },
  { section: 'VI. Ship of Theseus (Plutarch)', server: 'philosophy-mcp' },
  {
    section:
      'The Greek Logic Canon: Twelve Paradoxes and Mysteries (§20+§22+§23)',
    server: 'philosophy-mcp',
  },
  { section: "1. Zeno's Dichotomy", server: 'philosophy-mcp' },
  { section: '2. Achilles and the Tortoise', server: 'philosophy-mcp' },
  { section: "3. Zeno's Arrow", server: 'philosophy-mcp' },
  { section: "4. Meno's Paradox", server: 'philosophy-mcp' },
  { section: '5. The Sorites Paradox (Eubulides)', server: 'philosophy-mcp' },
  { section: '6. The Epicurean Swerve (Clinamen)', server: 'philosophy-mcp' },
  { section: '7. The Third Man Argument', server: 'philosophy-mcp' },
  { section: "8. Aristotle's Sea Battle", server: 'philosophy-mcp' },
  { section: "9. Aristotle's Golden Mean", server: 'philosophy-mcp' },
  { section: '10. Heraclitus vs Parmenides', server: 'philosophy-mcp' },
  { section: "11. Aristotle's Prime Mover", server: 'philosophy-mcp' },
  {
    section: '12. Unity of Virtues (Socrates/Protagoras)',
    server: 'philosophy-mcp',
  },
  {
    section:
      'Unsolved Mysteries: Fork/Race/Fold Perspectives (§17+§20+§22+§23)',
    server: 'philosophy-mcp',
  },
  {
    section: 'I. Fine-Tuning of Physics',
    server: 'science-mcp',
    coverageKind: 'umbrella',
  },
  { section: 'II. Baryon Asymmetry', server: 'quantum-mcp' },
  { section: 'III. Cambrian Explosion', server: 'genomics-mcp' },
  { section: 'IV. Evolution of Sex', server: 'genomics-mcp' },
  { section: 'V. Göbekli Tepe', server: 'anthropology-mcp' },
  { section: 'VI. Spontaneous Remission', server: 'genomics-mcp' },
  { section: 'VII. Origin of Language', server: 'language-mcp' },
  { section: 'VIII. Cosmological Constant', server: 'relativity-mcp' },
  {
    section: 'Second Tier Mysteries: How Did They Do That? (§17+§22+§23+§25)',
    server: 'philosophy-mcp',
  },
  {
    section: 'I. Megalithic Coordination (Sachsayhuamán, Nan Madol, Longyou)',
    server: 'anthropology-mcp',
  },
  {
    section: 'II. Undeciphered Scripts (Linear A, Rongorongo, Proto-Elamite)',
    server: 'language-mcp',
  },
  { section: 'III. The Mpemba Effect', server: 'thermodynamics-mcp' },
  { section: 'IV. Roman Dodecahedrons', server: 'history-mcp' },
  {
    section: 'V. Anomalous Evidence (Pollock Twins, Man from Taured, etc.)',
    server: 'evidence-mcp',
  },
  { section: 'VI. The Hum', server: 'science-mcp', coverageKind: 'umbrella' },
  {
    section:
      'Philosophical Combinatorics: Cross-Allegory Compositions (§15.30)',
    server: 'philosophy-mcp',
  },
  {
    section:
      'Philosophical Combinatorics Round 2: Philosophy × Science × Engineering (§15.30)',
    server: 'philosophy-mcp',
  },
  {
    section:
      'Philosophical Combinatorics Round 3: Stress-Testing the Limits (§15.30)',
    server: 'philosophy-mcp',
  },
  {
    section:
      'Philosophical Combinatorics Round 4: Seven Universal Laws (§15.30)',
    server: 'philosophy-mcp',
  },
  // ── Seven Laws Predictions (Sections 99-103) ────────────────────────
  {
    section: 'Seven Laws Predictions: Wild Novel Predictions (§15.30+§19)',
    server: 'predictions-mcp',
  },
  {
    section:
      'Seven Laws Predictions Round 2: Music, Dreams, Humor, Love, and Death (§15.30+§19)',
    server: 'predictions-mcp',
  },
  {
    section:
      'Seven Laws Predictions Round 3: Consciousness, Mastery, Teaching, and the Universe (§15.30+§19)',
    server: 'predictions-mcp',
  },
  {
    section:
      'Seven Laws Predictions Round 4: Game Theory, Economics, Ecology, Ethics (§15.30+§19)',
    server: 'predictions-mcp',
  },
  {
    section: 'Seven Laws Predictions Round 5: The Grand Finale (§15.30+§19)',
    server: 'predictions-mcp',
  },
  // ── Hard Compositions (Section 104) ──────────────────────────────────
  {
    section: 'Hard Compositions: Real Analysis Territory (§15.30)',
    server: 'ergodicity-mcp',
  },
  // ── God Formula & Reduction (Sections 105-111) ───────────────────────
  {
    section: 'Surface Reduction: The Minimal Generating Set (§15.30.3)',
    server: 'god-formula-mcp',
  },
  {
    section: 'Deep Reduction: From Seven Laws to One Axiom (§15.30.3)',
    server: 'god-formula-mcp',
  },
  {
    section: 'The Primator: The Final Reduction (§15.30.3)',
    server: 'god-formula-mcp',
  },
  {
    section: 'The Ceiling: What the Framework Cannot Prove (§15.30.3)',
    server: 'god-formula-mcp',
  },
  {
    section:
      'The Gain: The Habitable Zone Between Floor and Ceiling (§15.30.3)',
    server: 'god-formula-mcp',
  },
  {
    section: 'The Control Statistic: Validating the Sandwich (§15.30.5)',
    server: 'god-formula-mcp',
  },
  { section: 'The God Formula (§15.30)', server: 'god-formula-mcp' },
  { section: '112. Complement Semantics (§15.31)', server: 'emotion-mcp' },
  { section: '113. Sentience (§15.32)', server: 'emotion-mcp' },
  { section: '114. Emotion Taxonomy (§15.32.9)', server: 'emotion-mcp' },
  { section: '115. Name It To Tame It (§15.32.9)', server: 'emotion-mcp' },
  { section: '116. Emotion Theorems (§15.32.10)', server: 'emotion-mcp' },
  {
    section: '117. Emotion Cross-Compositions Round 1 (§15.32.10)',
    server: 'emotion-mcp',
  },
  {
    section: '118. Emotion Cross-Compositions Round 2 (§15.32.10)',
    server: 'emotion-mcp',
  },
  {
    section: '119. Emotion Cross-Compositions Round 3 (§15.32.10)',
    server: 'emotion-mcp',
  },
  {
    section: '120. Emotion Cross-Compositions Rounds 4-6 (§15.32.10)',
    server: 'emotion-mcp',
  },
  {
    section: '121. Emotion Cross-Compositions Rounds 7-13 (§15.32.10)',
    server: 'emotion-mcp',
  },
  {
    section: '122. Emotion Cross-Compositions Rounds 14-25 (§15.32.10)',
    server: 'emotion-mcp',
  },
  { section: '123. The Emotion Sandwich (§15.32.11)', server: 'emotion-mcp' },
  { section: 'PsycheGrind Promotion (2026-05-11)', server: 'emotion-mcp' },
  { section: 'Placebic Cognition Promotion (2026-06-16)', server: 'psyche-mcp' },
  {
    section: 'Sardis Distributed Intelligence Guard Promotion (2026-06-16)',
    server: 'metacognition-mcp',
  },
  { section: 'TopologicalMetabolism Promotion (2026-05-11)', server: 'thermodynamics-mcp' },
  // gnosis-math-mcp-placement:start
  { section: 'Gnosis Math Placement -- cybersecurity-mcp (2026-05-11)', server: 'cybersecurity-mcp' },
  { section: 'Gnosis Math Placement -- economics-mcp (2026-05-11)', server: 'economics-mcp' },
  { section: 'Gnosis Math Placement -- emotion-mcp (2026-05-11)', server: 'emotion-mcp' },
  { section: 'Gnosis Math Placement -- ergodicity-mcp (2026-05-11)', server: 'ergodicity-mcp' },
  { section: 'Gnosis Math Placement -- genomics-mcp (2026-05-11)', server: 'genomics-mcp' },
  { section: 'Gnosis Math Placement -- god-formula-mcp (2026-05-11)', server: 'god-formula-mcp' },
  { section: 'Gnosis Math Placement -- governance-mcp (2026-05-11)', server: 'governance-mcp' },
  { section: 'Gnosis Math Placement -- philosophy-mcp (2026-05-11)', server: 'philosophy-mcp' },
  { section: 'Gnosis Math Placement -- quantum-mcp (2026-05-11)', server: 'quantum-mcp' },
  { section: 'Gnosis Math Placement -- relativity-mcp (2026-05-11)', server: 'relativity-mcp' },
  { section: 'Gnosis Math Placement -- sleep-mcp (2026-05-11)', server: 'sleep-mcp' },
  { section: 'Gnosis Math Placement -- thermodynamics-mcp (2026-05-11)', server: 'thermodynamics-mcp' },
  { section: 'Gnosis Math Placement -- topology-mcp (2026-05-11)', server: 'topology-mcp' },
  { section: 'Gnosis Math Placement -- verification-mcp (2026-05-11)', server: 'verification-mcp' },
  // gnosis-math-mcp-placement:end
  // ── storms.watch Weather Manifold (§30-§32) ─────────────────────────
  { section: 'storms.watch State Reduction (§30)', server: 'storms-mcp' },
  { section: 'storms.watch Trident Weather Void (§30)', server: 'storms-mcp' },
  { section: 'storms.watch BATNA Shell Topology (§30)', server: 'storms-mcp' },
  { section: 'storms.watch Void Markov Property (§30)', server: 'storms-mcp' },
  {
    section: 'storms.watch Scoring Rule Properness (§30)',
    server: 'storms-mcp',
  },
  {
    section: 'storms.watch Thermodynamic Structure (§32)',
    server: 'storms-mcp',
  },
  {
    section: 'storms.watch Void Geometry and Metric (§32)',
    server: 'storms-mcp',
  },
  { section: 'storms.watch Algebraic Structure (§32)', server: 'storms-mcp' },
  { section: 'storms.watch Conservation Laws (§32)', server: 'storms-mcp' },
  {
    section: 'storms.watch Variational Principles (§32)',
    server: 'storms-mcp',
  },
  { section: 'storms.watch Number Theory (§32)', server: 'storms-mcp' },
  {
    section: 'storms.watch Scaling and Renormalization (§32)',
    server: 'storms-mcp',
  },
  { section: 'storms.watch Information Theory (§32)', server: 'storms-mcp' },
  { section: 'storms.watch Combinatorics (§32)', server: 'storms-mcp' },
  {
    section: 'storms.watch Extremal and Optimization (§32)',
    server: 'storms-mcp',
  },
  { section: 'storms.watch Category and Homology (§32)', server: 'storms-mcp' },
  {
    section: 'storms.watch Dynamics and Stability (§32)',
    server: 'storms-mcp',
  },
  {
    section: 'storms.watch Operational Forecasting (§32)',
    server: 'storms-mcp',
  },
  { section: 'storms.watch Game Theory (§32)', server: 'storms-mcp' },
  {
    section: 'storms.watch Probabilistic Structure (§32)',
    server: 'storms-mcp',
  },
  {
    section: 'storms.watch Demand and Competition (§31)',
    server: 'storms-mcp',
  },
  {
    section: 'storms.watch Void Algebra and Foundations (§30)',
    server: 'storms-mcp',
  },
  { section: 'storms.watch Trident Void Hardened (§30)', server: 'storms-mcp' },
  {
    section: 'storms.watch Source Quotients Hardened (§30)',
    server: 'storms-mcp',
  },
  {
    section: 'storms.watch Void Net Mass Sandwich (§30)',
    server: 'storms-mcp',
  },
  {
    section: 'storms.watch Residual Deficit Sandwich (§30)',
    server: 'storms-mcp',
  },
  { section: 'storms.watch Persistence Baseline (§30)', server: 'storms-mcp' },
  {
    section: 'storms.watch Skill as Void Contribution (§30)',
    server: 'storms-mcp',
  },
  { section: 'storms.watch Skill Sandwich (§30)', server: 'storms-mcp' },
  {
    section: 'storms.watch Forecast Score Sandwich (§30)',
    server: 'storms-mcp',
  },
  { section: 'storms.watch Score Monotonicity (§30)', server: 'storms-mcp' },
  {
    section: 'storms.watch Forcing Balance Sandwich (§30)',
    server: 'storms-mcp',
  },
  {
    section: 'storms.watch Viable Void Skill Guarantee (§30)',
    server: 'storms-mcp',
  },
  { section: 'storms.watch Shear Suppression (§30)', server: 'storms-mcp' },
  {
    section: 'storms.watch Pressure-Wind Consistency (§30)',
    server: 'storms-mcp',
  },
  {
    section: 'storms.watch Land Interaction Decay (§30)',
    server: 'storms-mcp',
  },
  {
    section: 'storms.watch Symmetry-Organization Coupling (§30)',
    server: 'storms-mcp',
  },
  {
    section: 'storms.watch Track Steering Alignment (§30)',
    server: 'storms-mcp',
  },
  { section: 'storms.watch Growth Tendency (§30)', server: 'storms-mcp' },
  { section: 'storms.watch Climatology Baseline (§30)', server: 'storms-mcp' },
  {
    section: 'storms.watch Persistence-Climatology Ordering (§30)',
    server: 'storms-mcp',
  },
  { section: 'storms.watch Score Decomposition (§30)', server: 'storms-mcp' },
  { section: 'storms.watch Target Bias Ordering (§30)', server: 'storms-mcp' },
  { section: 'storms.watch Two-Step Composition (§30)', server: 'storms-mcp' },
  {
    section: 'storms.watch Void Geometry Race Projections (§30)',
    server: 'storms-mcp',
  },
  {
    section: 'storms.watch Void Geometry Race Sandwiches (§30)',
    server: 'storms-mcp',
  },
  {
    section: 'storms.watch Void Geometry Race Selection (§30)',
    server: 'storms-mcp',
  },
  {
    section: 'storms.watch Void Geometry Fold Hierarchy (§30)',
    server: 'storms-mcp',
  },
  {
    section: 'storms.watch Impossibility Theorems (§30)',
    server: 'storms-mcp',
  },
  {
    section: 'storms.watch Weighted Void Geometry (§30)',
    server: 'storms-mcp',
  },
  {
    section: 'storms.watch Fork-Race-Fold Void Geometry (§30)',
    server: 'storms-mcp',
  },
  { section: 'storms.watch Sliver Guarantee (§30)', server: 'storms-mcp' },
  { section: 'storms.watch Reduction Theorem (§30)', server: 'storms-mcp' },
  { section: 'storms.watch Multi-Step Trajectory (§30)', server: 'storms-mcp' },
  { section: 'storms.watch Anti-Reduction (§30)', server: 'storms-mcp' },
  {
    section: 'storms.watch Convergence and Steady State (§30)',
    server: 'storms-mcp',
  },
  {
    section: 'storms.watch Divergence and Regime Change (§30)',
    server: 'storms-mcp',
  },
  { section: 'storms.watch Complete Picture (§30)', server: 'storms-mcp' },
  { section: 'storms.watch Extended Horizons (§30)', server: 'storms-mcp' },
  { section: 'storms.watch N-Step Iteration (§30)', server: 'storms-mcp' },
  { section: 'storms.watch Skill Decay (§30)', server: 'storms-mcp' },
  {
    section: 'storms.watch Convergence to Climatology (§30)',
    server: 'storms-mcp',
  },
  { section: 'storms.watch Weather Replay (§30)', server: 'storms-mcp' },
  {
    section: 'storms.watch Vulnerability and Bifurcation Dynamics (§30)',
    server: 'storms-mcp',
  },
  {
    section: 'storms.watch Path Action and God Weight Consolidation (§30)',
    server: 'storms-mcp',
  },
  // ── PsycheGrindExtended orphan passes ────────────────
  {
    section: 'Pass 20: Self-Organized Organism (from PsycheGrindExtended)',
    server: 'psyche-mcp',
  },
  {
    section: 'Pass 11: Digital Twins x Knot Theory (15 theorems)',
    server: 'psyche-mcp',
  },
  {
    section: 'Pass 12: Emotions x Void x Consciousness (18 theorems)',
    server: 'psyche-mcp',
  },
  {
    section: 'Pass 13: Language x Semiotic Deficit (16 theorems)',
    server: 'psyche-mcp',
  },
  { section: 'Pass 14: Cognitive Acumen (15 theorems)', server: 'psyche-mcp' },
  {
    section: 'Pass 15: Folded-Space Data Knots (12 theorems)',
    server: 'psyche-mcp',
  },
  { section: 'Pass 16: Convergence Master Theorem', server: 'psyche-mcp' },
  {
    section:
      'Pass 17: Self-Encoding Structures -- All Shapes of Frozen Pizza (20 theorems)',
    server: 'psyche-mcp',
  },
  {
    section: 'Pass 18: Void Merkle Verification (12 theorems)',
    server: 'psyche-mcp',
  },
  {
    section: 'Memory Leak as Counterfeit Liveness',
    server: 'psyche-mcp',
  },
  {
    section:
      'Pass 19: The Isomorphism -- FoldedKnot ≅ defenseWeight (15 theorems)',
    server: 'psyche-mcp',
  },
  // ── Scanner Theorem Surfaces ────────────────
  {
    section:
      'Scanner: Performance Obstruction Theory (PerformanceObstruction.lean)',
    server: 'polyglot-scanner',
  },
  {
    section: 'Scanner: Flow Obstruction Theory (FlowObstruction.lean)',
    server: 'polyglot-scanner',
  },
  {
    section: 'Scanner: Void Dimensions (VoidDimensions.lean)',
    server: 'polyglot-scanner',
  },
  {
    section: 'Scanner: Bug-Knot Correspondence (BugKnotCorrespondence.lean)',
    server: 'polyglot-scanner',
  },
  {
    section:
      'Scanner: String Theory Correspondence (StringTheoryCorrespondence.lean)',
    server: 'polyglot-scanner',
  },
  {
    section: 'Scanner: The Reduction (TheReduction.lean)',
    server: 'polyglot-scanner',
  },
  {
    section: 'Scanner: Empirical Detector Theorems (EmpiricalDetectors.lean)',
    server: 'polyglot-scanner',
  },
  {
    section:
      'Spiderweb BFT — Byzantine Fault Tolerance on Gnostic Meshes (2026-03-29)',
    server: 'topology-mcp',
  },
  {
    section: 'Spiderweb Mesh Network (2026-03-29)',
    server: 'topology-mcp',
  },
  {
    section: 'Spiderweb Topology (2026-03-29)',
    server: 'topology-mcp',
  },
  {
    section: 'Spiderweb Gnostic Ladder (2026-03-29)',
    server: 'topology-mcp',
  },
  {
    section: 'Valentinian Sin Diagnostic (2026-03-29)',
    server: 'topology-mcp',
  },
  {
    section: 'Valentinian Cross-Pollination (2026-03-29)',
    server: 'topology-mcp',
  },
  {
    section: 'Deception Topology (2026-03-27)',
    server: 'topology-mcp',
  },
  {
    section: 'Fighting Techniques x Combat Topology (2026-03-27)',
    server: 'topology-mcp',
  },
  {
    section: 'Combat Applications: Abstract Utility (2026-03-27)',
    server: 'topology-mcp',
  },
  {
    section: 'Mediation and Negotiation Topology (2026-03-27)',
    server: 'topology-mcp',
  },
  {
    section: 'Topological Contagion & Syzygy (2026-03-27)',
    server: 'topological-contagion-syzygy-mcp',
  },
  {
    section: 'Syzygy',
    server: 'topological-contagion-syzygy-mcp',
  },
  {
    section: 'Inductive Reidemeister Topology & The Hope Gap (2026-03-27)',
    server: 'topology-mcp',
  },
  {
    section: 'Liability Equilibrium (2026-03-29)',
    server: 'topology-mcp',
  },
  {
    section: 'Immigration Topology (2026-03-29, simplified)',
    server: 'topology-mcp',
  },
  {
    section: 'Immigration as Diversity Growth (2026-03-29, simplified)',
    server: 'topology-mcp',
  },
  {
    section:
      'Greedy Rejection & The Hope Gap — Deep Formalization (2026-03-29)',
    server: 'topology-mcp',
  },
  {
    section: 'Topological Compression & Information Dilation (2026-03-27)',
    server: 'topology-mcp',
  },
  {
    section: 'Pass 12: The Right to Exit (2026-03-27)',
    server: 'topology-mcp',
  },
  {
    section: 'Pass 64: Collective Sentience & Adaptive Synchrony (2026-03-27)',
    server: 'topology-mcp',
  },
  {
    section: 'Pass 60.6: The Meaning of Life — Untangling (2026-03-27)',
    server: 'topology-mcp',
  },
  {
    section: 'Pass 53: Machine Metabolism & Thermal Efficiency (2026-03-27)',
    server: 'topology-mcp',
  },
  {
    section: 'Pass 61: Inner Dialog — Speculative Cover Space (2026-03-27)',
    server: 'topology-mcp',
  },
  {
    section: 'Pass 62: Thought — Simplicial Collapse (2026-03-27)',
    server: 'topology-mcp',
  },
  {
    section:
      'Pass 65: The Hodge Bridge — Algebraic vs Topological Invariants (2026-03-27)',
    server: 'topology-mcp',
  },
  {
    section:
      'Pass 70: Aperiodic Scheduling & Wallington Resonance-Free Execution (2026-03-27)',
    server: 'topology-mcp',
  },
  {
    section: 'Kitchen & High-Intensity Communication Theory (2026-03-27)',
    server: 'topology-mcp',
  },
  {
    section: 'Humor Topology & Thermodynamic Venting (2026-04-01)',
    server: 'topology-mcp',
  },
  {
    section: 'Abstract Foundations',
    server: 'topology-mcp',
  },
  {
    section: 'Spiderweb Topology',
    server: 'topology-mcp',
  },
  {
    section: 'Spiderweb Topology (2026-03-29)',
    server: 'topology-mcp',
  },
  {
    section: 'Spiderweb Gnostic Ladder',
    server: 'topology-mcp',
  },
  {
    section: 'Spiderweb Gnostic Ladder (2026-03-29)',
    server: 'topology-mcp',
  },
  {
    section: 'Core Primitives',
    server: 'topology-mcp',
  },
  {
    section: 'Canonical Web Factorizations',
    server: 'topology-mcp',
  },
  {
    section: 'Cross-Level Structural Theorems',
    server: 'topology-mcp',
  },
  {
    section: 'Oracle Execution Stall Remedies (Douglasii)',
    server: 'verification-mcp',
  },
  {
    section: 'Moonshot Convergences (Douglasii)',
    server: 'verification-mcp',
  },
  {
    section: 'Iterative Cross-Pollination Pass (Cycle Douglasii - 5 Families)',
    server: 'verification-mcp',
  },
  {
    section: 'Cycle 7 Wild Exploration (2026-04-06)',
    server: 'verification-mcp',
  },
  {
    section:
      "Iterative Cross-Pollination Pass (Cycle 2026-04-06) (Buley's Run 3)",
    server: 'verification-mcp',
  },
  {
    section: 'Iterative Cross-Pollination Pass (Cycle 2026-04-06 Moonshots)',
    server: 'verification-mcp',
  },
  {
    section: 'Cycle 8 Wild Exploration (2026-04-06)',
    server: 'verification-mcp',
  },
  {
    section: 'Cycle 9 Iterative Pass (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section: 'Iterative Cross-Pollination Pass (Cycle 9 Iterative Pass 2)',
    server: 'verification-mcp',
  },
  {
    section: 'Cycle 9 Iterative Pass 2 (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section: 'Cycle 9 Iterative Pass 3 (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 8) (2026-04-07)',
    server: 'verification-mcp',
  },
  {
    section:
      'Iterative Cross-Pollination Pass (Cycle Douglasii Run 7) (2026-04-07)',
    server: 'verification-mcp',
  },
] as const;

export const THEOREM_OVERRIDES: Readonly<Record<string, OwnershipOverride>> = {
  'THM-VOID-ATTENTION-DECEPTACON-PAPER-LABEL': {
    server: 'metacognition-mcp',
    coverageKind: 'semantic',
    toolTheoremId: 'THM-VOID-ATTENTION',
  },
  'THM-BATNA-is-VOID-DECEPTACON-PAPER-LABEL': {
    server: 'metacognition-mcp',
    coverageKind: 'semantic',
    toolTheoremId: 'THM-BATNA-is-VOID',
  },
  'THM-DUAL-VOID-PARTITION-DECEPTACON-PAPER-LABEL': {
    server: 'metacognition-mcp',
    coverageKind: 'semantic',
    toolTheoremId: 'THM-DUAL-VOID-PARTITION',
  },
  'THM-WATNA-REDUCED-REGRET-DECEPTACON-PAPER-LABEL': {
    server: 'metacognition-mcp',
    coverageKind: 'semantic',
    toolTheoremId: 'THM-WATNA-REDUCED-REGRET',
  },
  'THM-COHERENCE-BREAKDOWN-DECEPTACON-PAPER-LABEL': {
    server: 'metacognition-mcp',
    coverageKind: 'semantic',
    toolTheoremId: 'THM-COHERENCE-BREAKDOWN',
  },
  /** CROSS_MIXED THX-R2-1 narrative row; mechanized core TBD. */
  'THM-XSYN-BEHAVIOR-EXTERNALIZATION': {
    server: 'communication-mcp',
  },
  'THM-RG-FIXED-POINTS': { server: 'coarsening-mcp' },
  'THM-THERMO-TRACED-MONOIDAL': { server: 'coherence-mcp' },
  'THM-FOLD-HEAT-HIERARCHY': { server: 'thermodynamics-mcp' },
  'THM-SEMIOTIC-PEACE': { server: 'conflict-mcp' },
  'THM-COMMUNITY-ATTENUATES-FAILURE': { server: 'coarsening-mcp' },
  'THM-COMMUNITY-MONOTONE-NONDEGRADATION': { server: 'coarsening-mcp' },
  'THM-COMMUNITY-STRICT-DOMINATION': { server: 'coarsening-mcp' },
  'THM-COMMUNITY-BRIDGES-TARES': { server: 'coarsening-mcp' },
  'THM-COMMUNITY-BULE-CONVERGENCE': { server: 'coarsening-mcp' },
  'THM-COMMUNITY-DOMINANCE-THEORY': { server: 'coarsening-mcp' },
  'THM-COMMUNITY-DOMINANCE-MODEL': { server: 'coarsening-mcp' },
  'THM-COMMUNITY-is-SKYRMS-WALKER': { server: 'coarsening-mcp' },
  'THM-BULE-ZERO-is-NADIR': { server: 'coarsening-mcp' },
  'THM-NADIR-ALGEBRAIC': { server: 'coarsening-mcp' },
  'THM-SKYRMS-NADIR-is-BULE-ZERO': { server: 'coarsening-mcp' },
  'THM-SKYRMS-NADIR-BULE-MODEL': { server: 'coarsening-mcp' },
  'THM-DIVERSITY-OPTIMALITY': { server: 'coarsening-mcp' },
  'THM-ALLEGORIES-MASTER': {
    server: 'philosophy-mcp',
    coverageKind: 'umbrella',
  },
  'THM-GREEK-CANON-MASTER': {
    server: 'philosophy-mcp',
    coverageKind: 'umbrella',
  },
  'THM-MYSTERIES-MASTER': { server: 'science-mcp', coverageKind: 'umbrella' },
  'THM-SECOND-TIER-MASTER': {
    server: 'civilization-mcp',
    coverageKind: 'umbrella',
  },
  /** CacheObservation.lean marginal-fold model; queueing MCP hosts the numeric witness. */
  'THM-CALLBACK-RESONANCE': {
    server: 'queueing-mcp',
    coverageKind: 'wrapper',
  },
  // Section 123 Emotion Sandwich -- renamed from generic CONTROL IDs to EMOTION-prefixed
  // No overrides needed: IDs are now unique (THM-EMOTION-CONTROL-* vs THM-CONTROL-*)
  'THM-MOONSHOT-SEMANTIC-RESONANCE': { server: 'philosophy-mcp' },
  'THM-CROSS-CRYPTOGRAPHY-MYCOLOGY-BYPASS': { server: 'cybersecurity-mcp' },
  'THM-MOONSHOT-INTERPRETATION-TOPOLOGY-BYPASS': { server: 'topology-mcp' },
  'THM-MOONSHOT-WITNESS-GAP-HOMOLOGICAL-FUNCTOR': { server: 'mathematics-mcp' },
  'THM-MOONSHOT-INTERPRETATION-LAYER-TENSEGRITY-HOLOGRAPHY': { server: 'topology-mcp' },
  'THM-CROSS-OCEANOGRAPHY-ORACLE-STALL-ECONOMICS': { server: 'economics-mcp' },
  'THM-CROSS-LINGUISTICS-GASTRONOMY-WITNESS-GAP': { server: 'language-mcp' },
  'THM-MOONSHOT-HOMOTOPY-WITNESS-GAP': { server: 'mathematics-mcp' },
  'THM-CONTRARIAN-INTERPRETATIONLESS-ENTROPY': { server: 'thermodynamics-mcp' },
  'THM-BRIDGE-OCEANOGRAPHY-CRYPTO': { server: 'cybersecurity-mcp' },
  'THM-BRIDGE-BOTANY-ECONOMICS': { server: 'economics-mcp' },
  'THM-MOONSHOT-AESTHETIC-COMPRESSION': { server: 'design-mcp' },
  'THM-MOONSHOT-QUANTUM-CONTRACT': { server: 'quantum-mcp' },
  'THM-CONTRARIAN-OBSERVER-EFFECT': { server: 'quantum-mcp' },
  'THM-BRIDGE-MUSIC-ASTROPHYSICS': { server: 'astronomy-mcp' },
  'THM-BRIDGE-EPI-CODE-CHURN': { server: 'science-mcp' },
  'THM-BRIDGE-HOHMANN-MIGRATION': { server: 'astronomy-mcp' },
  'THM-CONTRARIAN-STRICT-TYPING': { server: 'cybersecurity-mcp' },
  'THM-CONTRARIAN-SIMPLICITY-FRAGILITY': { server: 'cybersecurity-mcp' },
  'THM-QUEUE-TECTONIC-SUBDUCTION-NOT-POSITIVE-BETA1': { server: 'science-mcp' },
  'THM-TECTONIC-TENSOR-GEOMETRIC-RATE': { server: 'science-mcp' },
  'THM-TECTONIC-TENSOR-SEMANTIC-MORPHISM-YIELDS-POSITIVE-TOPOLOGICAL-DEFICIT': {
    server: 'mathematics-mcp',
  },
  'THM-BRIDGE-GASTRONOMY-CRYPTO': { server: 'cybersecurity-mcp' },
  'THM-CONTRARIAN-OVERFITTING-GENERALIZES': { server: 'cybersecurity-mcp' },
  'THM-MOONSHOT-ESCHATOLOGICAL-CRYPTO': { server: 'cybersecurity-mcp' },
  'THM-ERGODIC-SEMANTIC-RESOLUTION': { server: 'verification-mcp' },
  'THM-FLUID-ECONOMICS-BRIDGE': { server: 'economics-mcp' },
  'THM-LINGUISTICS-THERMO-BRIDGE': { server: 'thermodynamics-mcp' },
  'THM-ORACLE-STALL-RESOLUTION': { server: 'verification-mcp' },
  'THM-PROOF-DEPTH-FRAGILITY': { server: 'verification-mcp' },
  'THM-SEMANTIC-TOPOLOGY-DRIFT': { server: 'topology-mcp' },
  'THM-STALL-is-OPTIMAL': { server: 'verification-mcp' },
  'THM-MOONSHOT-SUBATOMIC-BUREAUCRACY': { server: 'science-mcp' },
  'THM-MOONSHOT-ERGODIC-JAZZ': { server: 'ergodicity-mcp' },
  'THM-CONTRARIAN-LATENCY-KNOWLEDGE': { server: 'metacognition-mcp' },
  'THM-CROSS-DOMAIN-ASTRONOMY-CULINARY': { server: 'astronomy-mcp' },
  'THM-ORACLE-STALL-ANNIHILATION': { server: 'verification-mcp' },
  'THM-MOONSHOT-RECURSIVE-HOLOGRAPHIC-VERIFICATION': {
    server: 'verification-mcp',
  },
  'THM-MOONSHOT-TOPOLOGICAL-DEFICIT-COMPRESSION': { server: 'mathematics-mcp' },
  'THM-CONTRARIAN-PERFECT-INFORMATION-INEFFICIENCY': {
    server: 'thermodynamics-mcp',
  },
  'THM-CROSS-DOMAIN-TECTONIC-QUEUE-REORDER': { server: 'science-mcp' },
  'THM-CROSS-DOMAIN-MYCORRHIZAL-ZKP-VERIFICATION': {
    server: 'cybersecurity-mcp',
  },
  'THM-CROSS-DOMAIN-ORNITHOLOGY-CRYPTOGRAPHY': { server: 'cybersecurity-mcp' },
  'THM-CROSS-DOMAIN-METEOROLOGY-BLOCKCHAIN': { server: 'science-mcp' },
  'THM-MOONSHOT-HYPERBOLIC-CONSENSUS': { server: 'verification-mcp' },
  'THM-MOONSHOT-INFINITESIMAL-PREFILL': { server: 'verification-mcp' },
  'THM-CONTRARIAN-PRECISION-DEFICIT': { server: 'verification-mcp' },
  'THM-BRIDGE-MYCELIAL-ARCHITECTURE': { server: 'design-mcp' },
  'THM-CONTRARIAN-REDUNDANCY-FRAGILITY': { server: 'verification-mcp' },
  'THM-QUEUE-KATA-ZECKENDORF-BUDGET-UNIT-BOUNDARY': { server: 'queueing-mcp' },
  'THM-QUEUE-KATA-ZECKENDORF-JOINT-WITNESS-PACK': { server: 'queueing-mcp' },
  'THM-QUEUE-KATA-ZECKENDORF-BUDGET-NOT-POSITIVE-BETA1': {
    server: 'queueing-mcp',
  },
  'THM-QUEUE-KATA-ZECKENDORF-BUDGET-NOT-STRICT-CAPACITY-GROWTH': {
    server: 'queueing-mcp',
  },
  'THM-KATA-ZECKENDORF-GEOMETRIC-RATE-CERTIFICATE': {
    server: 'ergodicity-mcp',
  },
  'THM-KATA-ZECKENDORF-KERNEL-LIFT-BUDGET-POSITIVE-FROM-SOURCE': {
    server: 'ergodicity-mcp',
  },
  'THM-KATA-ZECKENDORF-KERNEL-LIFT-CONTINUOUS-ERGODICITY': {
    server: 'ergodicity-mcp',
  },
  'THM-KATA-ZECKENDORF-GAP-HYPOTHESIS-FROM-SOURCE': {
    server: 'mathematics-mcp',
  },
  'THM-KATA-ZECKENDORF-MULTILEVEL-HARRIS-WITNESS-VIA-GAP': {
    server: 'ergodicity-mcp',
  },
  'THM-KATA-ZECKENDORF-ADAPTER-RATE-SUBUNIT': { server: 'ergodicity-mcp' },
  'THM-KATA-ZECKENDORF-ADAPTER-STRICT-GEOMETRIC-DECAY': {
    server: 'ergodicity-mcp',
  },
  'THM-QUEUE-KATA-ZECKENDORF-MAJORITY-RATE-EMBEDDING-BOUNDARY': {
    server: 'queueing-mcp',
  },
  'THM-QUEUE-KATA-ZECKENDORF-NOT-SUBUNIT-RESIDENCE-TIME': {
    server: 'queueing-mcp',
  },
  'ANTI-THM-FULL-RECONSTRUCTION-VIOLATES-THERMO': {
    server: 'thermodynamics-mcp',
  },
  'ANTI-THM-STALL-FREE-NO-ENTANGLEMENT': {
    server: 'verification-mcp',
  },
  'ANTI-THM-STALLED-QUEUE-SELF-ORGANIZATION-REQUIRES-BUDGET': { server: 'queueing-mcp' },
  'THM-STALLED-QUEUE-SELF-ORGANIZATION-REQUIRES-BUDGET': { server: 'queueing-mcp' },
  'THM-MOONSHOT-TENSOR-FOLD-INVERSION': { server: 'mathematics-mcp' },
  'THM-MOONSHOT-DIRECT-ENTANGLED-KERNEL-LIFTING': { server: 'quantum-mcp' },
  'THM-CROSS-CYBERSECURITY-THERMODYNAMICS-PERMUTATION-ENTROPY': { server: 'cybersecurity-mcp' },
  'THM-OBSERVER-FREE-STATE-DIVERGENCE': { server: 'quantum-mcp' },
  'ANTI-THM-PALEOCLIMATOLOGY-SEISMOLOGY-BUDGET-NOT-POSITIVE-BETA1': { server: 'science-mcp' },
  'THM-MOONSHOT-INTERPRETATION-LAYER-RESONANCE-DUALITY': { server: 'verification-mcp' },
  'THM-MOONSHOT-WITNESS-GAP-TOPOLOGICAL-BRAID': { server: 'topology-mcp' },
  'THM-MOONSHOT-ORACLE-STALL-ASYMPTOTIC-FREEDOM': { server: 'verification-mcp' },
  'ANTI-THM-CONTRARIAN-INTERPRETATION-ABSENCE-INDUCES-SPEED': { server: 'verification-mcp' },
  'ANTI-THM-CONTRARIAN-ORACLE-STALL-PREVENTS-COLLAPSE': { server: 'thermodynamics-mcp' },
  'THM-CROSS-DOMAIN-ASTROPHYSICS-NEUROLOGY-ORACLE-BYPASS': { server: 'science-mcp' },
  'THM-CROSS-DOMAIN-MYCOLOGY-CRYPTOGRAPHY-WITNESS-BRIDGE': { server: 'cybersecurity-mcp' },
  'THM-PALEOCLIMATOLOGY-SEISMOLOGY-BUDGET-NOT-POSITIVE-BETA1': { server: 'science-mcp' },
};
