// @a0n/.ignored_gnosis/scripts/validate-ledger-mcp-coverage.ts
import assert from 'node:assert/strict';
import { existsSync, readFileSync } from 'node:fs';
import path from 'node:path';
import {
  buildGeneratedCoverage,
  EXPECTED_CANONICAL_LEDGER_ROW_COUNT,
  EXPECTED_CANONICAL_LEDGER_SECTION_COUNT,
  normalizeSectionName,
  resolveCanonicalLedgerPath,
  resolveProofArtifactPath,
  toolNameFromTheoremId,
  verifyGeneratedCoverageHasMcpHomes,
  type GeneratedCoverage,
} from './generate-ledger-mcp-coverage';

function assertGeneratedCounts(generated: GeneratedCoverage): void {
  assert.equal(generated.rows.length, EXPECTED_CANONICAL_LEDGER_ROW_COUNT);
  assert.equal(
    generated.audit.sectionCount,
    EXPECTED_CANONICAL_LEDGER_SECTION_COUNT
  );
  assert.equal(generated.audit.entries.length, generated.rows.length);
}

function assertNormalizationSamples(): void {
  const normalizedRaceWinner = normalizeSectionName(
    'Race Winner Correctness (Track Omicron)'
  );
  const normalizedPersonality = normalizeSectionName(
    'Five-Parameter Void Walker Personality Model (ch17 section 15.10.7, 15.12)'
  );
  const normalizedSelfHosting = normalizeSectionName(
    '72. Self-Hosting Optimality (§10.6)'
  );

  assert.equal(normalizedRaceWinner, 'Race Winner Correctness');
  assert.equal(
    normalizedPersonality,
    'Five-Parameter Void Walker Personality Model'
  );
  assert.equal(normalizedSelfHosting, 'Self-Hosting Optimality');
}

function assertCoverageFlags(generated: GeneratedCoverage): void {
  let missingCoverageCount = 0;
  let governanceCovered = false;
  let economicsCovered = false;

  for (const entry of generated.audit.entries) {
    if (!entry.server || !entry.tool || !entry.coverageKind) {
      missingCoverageCount += 1;
    }
    if (entry.server === 'governance-mcp') {
      governanceCovered = true;
    }
    if (entry.server === 'economics-mcp') {
      economicsCovered = true;
    }
  }

  assert.equal(missingCoverageCount, 0);
  assert.ok(governanceCovered);
  assert.ok(economicsCovered);
}

function assertDuplicateTheoremFidelity(generated: GeneratedCoverage): void {
  for (const theoremId of generated.audit.duplicateTheoremIds) {
    const duplicateEntries = generated.audit.entries.filter(
      (entry) => entry.theoremId === theoremId
    );
    const owners = new Set(duplicateEntries.map((entry) => entry.server));
    const tools = new Set(duplicateEntries.map((entry) => entry.tool));

    assert.equal(
      owners.size,
      1,
      `Duplicate theorem ${theoremId} maps to multiple MCP owners.`
    );
    assert.equal(
      tools.size,
      1,
      `Duplicate theorem ${theoremId} maps to multiple MCP tool names.`
    );
  }
}

function assertContractFidelity(generated: GeneratedCoverage): void {
  const rowsByIndex = new Map(
    generated.rows.map((row) => [row.rowIndex, row] as const)
  );

  for (const entry of generated.audit.entries) {
    const sourceRow = rowsByIndex.get(entry.rowIndex);
    assert.ok(sourceRow, `Missing source row for audit entry ${entry.rowIndex}.`);
    assert.equal(entry.assumptions, sourceRow.assumptions);
    assert.equal(
      entry.proofArtifactPath,
      resolveProofArtifactPath(entry.toolTheoremId, sourceRow.mechanization)
    );
    assert.ok(
      existsSync(path.join(process.cwd(), entry.proofArtifactPath)),
      `Missing proof artifact at ${entry.proofArtifactPath}.`
    );
  }
}

function assertSemanticAliases(generated: GeneratedCoverage): void {
  const entriesByTheoremId = new Map(
    generated.audit.entries.map((entry) => [entry.theoremId, entry])
  );

  for (const entry of generated.audit.entries) {
    if (entry.coverageKind !== 'semantic') {
      continue;
    }

    assert.ok(entry.theoremId !== entry.toolTheoremId);
    assert.equal(entry.tool, toolNameFromTheoremId(entry.toolTheoremId));

    const target = entriesByTheoremId.get(entry.toolTheoremId);
    assert.ok(target, `Missing semantic alias target for ${entry.theoremId}.`);
    assert.equal(entry.tool, target.tool);
    assert.ok(target.coverageKind !== 'semantic');
  }
}

/**
 * Runs the gnosis command-line workflow.
 */
export async function main(): Promise<void> {
  const ledgerPath = resolveCanonicalLedgerPath();
  const source = readFileSync(ledgerPath, 'utf8');
  const generated = buildGeneratedCoverage(source);

  assertGeneratedCounts(generated);
  assertNormalizationSamples();
  assertCoverageFlags(generated);
  assertDuplicateTheoremFidelity(generated);
  assertContractFidelity(generated);
  assertSemanticAliases(generated);
  await verifyGeneratedCoverageHasMcpHomes(generated);
  return;
}
