// @a0n/.ignored_gnosis/rtlsdr-mock-sim/bpsk-modem.ts
/**
 * Coherent BPSK modem for the cache-key frame -- the power-efficient alternative
 * to the non-coherent M-FSK PHY (mfsk-modem.ts). BPSK needs ~4 dB less Eb/N0 at a
 * fixed BER but pays for it with carrier-phase recovery: each coded bit rides one
 * symbol as +amp (phase 0) or -amp (phase pi) on a carrier at baseFreqHz.
 *
 * Frame: [PREAMBLE +/- alternations][SYNC bits][HEADER: uint16 blockLen, triplicated]
 *        [BODY: FEC(payload+crc16)], one coded bit per symbol.
 *
 * The preamble (a +1,-1,+1,-1 ... pattern, i.e. a tone at fc + baud/2) trains the
 * residual-CFO estimate and the carrier phase reference; the sync word resolves
 * the BPSK 0/pi ambiguity and the symbol boundary. Detection is a coherent matched
 * filter: de-rotate the baseband by the (CFO-corrected) carrier, integrate over the
 * sps samples of a symbol, and take the sign. Same FEC stack as MFSK, same frame
 * sizing math (one bit per symbol vs bps bits per symbol).
 *
 * This module is intentionally self-contained -- it does not import mfsk-modem.ts.
 */
import { complexFromU8Pair, u8FromComplexSample } from './iq';
import { rotateIqFrequency } from './pipeline';
import {
  codedBitLength,
  decodeFec,
  encodeFec,
  type FecProfile,
} from './fec-codec';

export interface BpskOptions {
  readonly iqRateHz: number;
  readonly baud: number;
  readonly baseFreqHz: number;
  readonly fec: FecProfile;
  /** Carrier amplitude in [-1,1] u8 scale. Encodes Eb/N0 vs the channel noise fill. */
  readonly amplitude?: number;
}

export const DEFAULT_BPSK: Omit<BpskOptions, 'fec'> = {
  iqRateHz: 2_400_000, // valid for both Pluto TX (min ~2.083 MS/s) and NESDR RX
  baud: 8000, // sps = 300 (integer)
  baseFreqHz: 96_000, // multiple of baud -> integer cycles/symbol, phase aligns on boundaries
  amplitude: 0.6,
};

// Alternating +/- preamble for timing + CFO + phase training (a tone at fc+baud/2).
// One BPSK symbol per entry; longer = a tighter CFO estimate.
const PREAMBLE_SYMS = 32;
// Known sync bits (BPSK +1/-1). Resolves the pi-ambiguity and the symbol boundary.
// A maximal-length-ish run with good autocorrelation; long enough to be distinctive.
const SYNC_BITS = [
  1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 1,
];
const HEADER_BITS = 16;
const HEADER_REPEAT = 3;

function spsForOpts(o: { iqRateHz: number; baud: number }): number {
  const sps = Math.round(o.iqRateHz / o.baud);
  if (sps < 8) throw new Error(`bpsk: samplesPerSymbol ${sps} too small`);
  return sps;
}

/** CRC-16/CCITT-FALSE over bytes (poly 0x1021, init 0xFFFF). Mirrors mfsk-modem. */
export function crc16Bytes(bytes: Uint8Array): number {
  let crc = 0xffff;
  for (let i = 0; i < bytes.length; i++) {
    crc ^= bytes[i]! << 8;
    for (let b = 0; b < 8; b++) {
      crc = (crc & 0x8000) !== 0 ? ((crc << 1) ^ 0x1021) & 0xffff : (crc << 1) & 0xffff;
    }
  }
  return crc & 0xffff;
}

/** uint16 blockLen, triplicated, MSB-first -> a bit array of length HEADER_BITS*HEADER_REPEAT. */
function headerBits(blockLen: number): Uint8Array {
  const bits = new Uint8Array(HEADER_BITS * HEADER_REPEAT);
  for (let r = 0; r < HEADER_REPEAT; r++) {
    for (let b = 0; b < HEADER_BITS; b++) {
      bits[r * HEADER_BITS + b] = (blockLen >> (HEADER_BITS - 1 - b)) & 1;
    }
  }
  return bits;
}

/** Total BPSK symbols (= coded bits, one per symbol) before the preamble. */
function payloadSymCount(blockLen: number, fec: FecProfile): number {
  return SYNC_BITS.length + HEADER_BITS * HEADER_REPEAT + codedBitLength(blockLen, fec);
}

/** Total IQ samples (pairs) a frame occupies for a given payload length. */
export function frameSampleCount(payloadLen: number, o: BpskOptions): number {
  const sps = spsForOpts(o);
  const blockLen = payloadLen + 2; // + crc16
  const totalSyms = PREAMBLE_SYMS + payloadSymCount(blockLen, o.fec);
  return totalSyms * sps;
}

/**
 * Synthesize `sps` samples of a BPSK symbol (bit 1 -> +amp, bit 0 -> -amp) on the
 * carrier, phase-continuous via the running carrier phase. With integer
 * cycles/symbol the carrier phase returns to its boundary value each symbol, so
 * the BPSK +/- modulation is a clean per-symbol sign flip.
 */
function synthSymbol(
  out: number[],
  bit: number,
  sps: number,
  fc: number,
  fs: number,
  phase0: number,
  amp: number,
): number {
  const w = (Math.PI * 2 * fc) / fs;
  const sign = bit ? 1 : -1;
  let ph = phase0;
  for (let n = 0; n < sps; n++) {
    const re = sign * amp * Math.cos(ph);
    const im = sign * amp * Math.sin(ph);
    out.push(...u8FromComplexSample(re, im));
    ph += w;
  }
  return ph;
}

/**
 * Handles the gnosis modulate Bpsk Frame workflow.
 */
export function modulateBpskFrame(payload: Uint8Array, o: BpskOptions): Uint8Array {
  const sps = spsForOpts(o);
  const amp = o.amplitude ?? 0.6;
  // body = payload + crc16
  const crc = crc16Bytes(payload);
  const body = new Uint8Array(payload.length + 2);
  body.set(payload);
  body[payload.length] = (crc >> 8) & 0xff;
  body[payload.length + 1] = crc & 0xff;
  const blockLen = body.length;
  const coded = encodeFec(body, o.fec);

  // assemble the full bit stream after the preamble
  const hdr = headerBits(blockLen);
  const bits = new Uint8Array(SYNC_BITS.length + hdr.length + coded.length);
  let off = 0;
  for (const b of SYNC_BITS) bits[off++] = b;
  bits.set(hdr, off);
  off += hdr.length;
  bits.set(coded, off);

  const out: number[] = [];
  let phase = 0;
  // preamble: alternating 1,0,1,0,... (a clean +/- tone for acquisition)
  for (let k = 0; k < PREAMBLE_SYMS; k++) {
    phase = synthSymbol(out, k % 2 === 0 ? 1 : 0, sps, o.baseFreqHz, o.iqRateHz, phase, amp);
  }
  for (let k = 0; k < bits.length; k++) {
    phase = synthSymbol(out, bits[k]!, sps, o.baseFreqHz, o.iqRateHz, phase, amp);
  }
  return Uint8Array.from(out);
}

export interface BpskDemodResult {
  readonly ok: boolean;
  readonly payload: Uint8Array | null;
  readonly blockLen: number;
  readonly syncOffset: number;
  readonly syncScore: number;
  readonly rsErrors: number;
  readonly cfoHz: number;
  readonly reason?: string;
}

/**
 * Coherent matched-filter accumulator over one symbol: de-rotate the baseband by
 * the carrier and integrate. Returns the complex symbol value (re,im); the sign of
 * the real part (after the phase reference is applied) is the recovered bit.
 */
function symbolCorrelate(
  re: Float64Array,
  im: Float64Array,
  start: number,
  sps: number,
  fc: number,
  fs: number,
): { re: number; im: number } {
  const w = (Math.PI * 2 * fc) / fs;
  let cr = 0;
  let ci = 0;
  for (let n = 0; n < sps; n++) {
    const c = Math.cos(w * n);
    const s = Math.sin(w * n);
    const xr = re[start + n]!;
    const xi = im[start + n]!;
    // multiply by conj(carrier) = (c, -s): de-rotate to baseband
    cr += xr * c + xi * s;
    ci += xi * c - xr * s;
  }
  return { re: cr, im: ci };
}

/**
 * Preamble correlation score at a candidate offset. The preamble is +,-,+,- ...
 * so adjacent de-rotated symbols should be anti-correlated and equal in magnitude.
 * Returns a normalized [0,1]-ish score plus the average complex symbol direction
 * (the carrier phase reference) and an estimate of residual CFO from the preamble's
 * symbol-to-symbol phase progression.
 */
function preambleScore(
  re: Float64Array,
  im: Float64Array,
  offset: number,
  sps: number,
  fc: number,
  fs: number,
): { score: number; refRe: number; refIm: number; cfoRadPerSym: number } | null {
  if (offset + PREAMBLE_SYMS * sps > re.length) return null;
  // Collapse each preamble symbol; flip the odd ones so all should align.
  let accRe = 0;
  let accIm = 0;
  let energy = 0;
  let coherent = 0;
  const syms: Array<{ re: number; im: number }> = [];
  for (let k = 0; k < PREAMBLE_SYMS; k++) {
    const sv = symbolCorrelate(re, im, offset + k * sps, sps, fc, fs);
    const sign = k % 2 === 0 ? 1 : -1;
    const fre = sign * sv.re;
    const fim = sign * sv.im;
    accRe += fre;
    accIm += fim;
    energy += sv.re * sv.re + sv.im * sv.im;
    syms.push({ re: fre, im: fim });
  }
  if (energy <= 0) return null;
  const mag = Math.sqrt(accRe * accRe + accIm * accIm);
  // coherent / non-coherent energy ratio -> alignment quality
  for (const s of syms) coherent += Math.sqrt(s.re * s.re + s.im * s.im);
  const score = coherent > 0 ? mag / coherent : 0;

  // CFO from the NET phase across the preamble (two-half difference), not a per-step
  // average -- this is a much lower-variance estimate. Pure-noise inputs (true CFO=0)
  // average toward zero, and a deadband (applied by the caller) ignores the residual.
  const half = syms.length >> 1;
  let aRe = 0, aIm = 0, bRe = 0, bIm = 0;
  for (let k = 0; k < half; k++) { aRe += syms[k]!.re; aIm += syms[k]!.im; }
  for (let k = half; k < syms.length; k++) { bRe += syms[k]!.re; bIm += syms[k]!.im; }
  // phase of (sum of 2nd half) * conj(sum of 1st half), spread over `half` symbols
  const dr = bRe * aRe + bIm * aIm;
  const di = bIm * aRe - bRe * aIm;
  const netPhase = dr * dr + di * di > 0 ? Math.atan2(di, dr) : 0;
  const cfoRadPerSym = half > 0 ? netPhase / half : 0;
  return { score, refRe: accRe / mag, refIm: accIm / mag, cfoRadPerSym };
}

/**
 * Handles the gnosis demodulate Bpsk Frame workflow.
 */
export function demodulateBpskFrame(
  iqU8: Uint8Array,
  o: BpskOptions,
  searchFromSample = 0,
  maxSearchSamples = Number.POSITIVE_INFINITY,
): BpskDemodResult {
  const sps = spsForOpts(o);
  const nPairs = Math.floor(iqU8.length / 2);
  const pad = sps; // zero-pad tail so a few-sample sync offset never truncates
  const re = new Float64Array(nPairs + pad);
  const im = new Float64Array(nPairs + pad);
  for (let p = 0; p < nPairs; p++) {
    const c = complexFromU8Pair(iqU8[2 * p]!, iqU8[2 * p + 1]!);
    re[p] = c.re;
    im[p] = c.im;
  }

  // ---- coarse + fine acquisition on the preamble (early-accept a strong peak) ----
  const minSyms = PREAMBLE_SYMS + SYNC_BITS.length + HEADER_BITS * HEADER_REPEAT + 1;
  const maxStart = Math.min(
    Math.max(0, nPairs - minSyms * sps),
    searchFromSample + maxSearchSamples,
  );
  const coarse = Math.max(4, sps >> 3);
  const EARLY_ACCEPT = 0.9;
  let bestOff = -1;
  let bestScore = -1;
  for (let off = searchFromSample; off <= maxStart; off += coarse) {
    const ps = preambleScore(re, im, off, sps, o.baseFreqHz, o.iqRateHz);
    if (ps && ps.score > bestScore) {
      bestScore = ps.score;
      bestOff = off;
    }
    if (ps && ps.score >= EARLY_ACCEPT) break;
  }
  if (bestOff < 0) {
    return emptyFail(bestScore, 'no preamble');
  }
  // fine search +/- a coarse step at single-sample resolution
  for (let off = Math.max(0, bestOff - coarse); off <= bestOff + coarse; off++) {
    if (off > maxStart) break;
    const ps = preambleScore(re, im, off, sps, o.baseFreqHz, o.iqRateHz);
    if (ps && ps.score > bestScore) {
      bestScore = ps.score;
      bestOff = off;
    }
  }

  // ---- residual CFO correction from the preamble, then re-derive phase ref ----
  // Only de-rotate when the estimated CFO would drift the phase appreciably across
  // the frame. The preamble CFO estimate has noise; applying a pure-noise CFO (true
  // CFO=0 on this bench) corrupts the body, so a deadband keeps small estimates off.
  // Deadband: ignore drift smaller than ~6 deg/symbol-over-the-whole-frame budget,
  // i.e. CFO whose per-symbol phase is below CFO_DEADBAND_RAD.
  const CFO_DEADBAND_RAD = (Math.PI * 2) / 256; // ~1.4 deg/symbol
  let workRe = re;
  let workIm = im;
  let baseOff = bestOff;
  let cfoHz = 0;
  const ps0 = preambleScore(re, im, bestOff, sps, o.baseFreqHz, o.iqRateHz)!;
  if (Math.abs(ps0.cfoRadPerSym) > CFO_DEADBAND_RAD) {
    cfoHz = (ps0.cfoRadPerSym * o.baud) / (Math.PI * 2);
    const rotated = rotateIqFrequency(iqU8, o.iqRateHz, -cfoHz);
    const rPairs = Math.floor(rotated.length / 2);
    const rre = new Float64Array(rPairs + pad);
    const rim = new Float64Array(rPairs + pad);
    for (let p = 0; p < rPairs; p++) {
      const c = complexFromU8Pair(rotated[2 * p]!, rotated[2 * p + 1]!);
      rre[p] = c.re;
      rim[p] = c.im;
    }
    workRe = rre;
    workIm = rim;
  }

  // phase reference: average direction of the (sign-flipped) preamble symbols on
  // the CFO-corrected stream.
  const psRef = preambleScore(workRe, workIm, baseOff, sps, o.baseFreqHz, o.iqRateHz);
  if (!psRef) return emptyFail(bestScore, 'preamble ref lost');
  let refRe = psRef.refRe;
  let refIm = psRef.refIm;

  // ---- resolve the pi-ambiguity + confirm boundary with the sync word ----
  const dataStart = baseOff + PREAMBLE_SYMS * sps;
  // project the sync symbols onto the phase reference; the preamble fixes the global
  // phase up to a 0/pi flip, which the sync correlation resolves.
  const syncSoft = new Float64Array(SYNC_BITS.length);
  for (let k = 0; k < SYNC_BITS.length; k++) {
    const sv = symbolCorrelate(workRe, workIm, dataStart + k * sps, sps, o.baseFreqHz, o.iqRateHz);
    syncSoft[k] = sv.re * refRe + sv.im * refIm; // real projection on the ref axis
  }
  let corr = 0;
  for (let k = 0; k < SYNC_BITS.length; k++) {
    corr += (SYNC_BITS[k] ? 1 : -1) * syncSoft[k]!;
  }
  let polarity = 1;
  if (corr < 0) {
    // 180-degree flip: invert the phase reference
    polarity = -1;
    refRe = -refRe;
    refIm = -refIm;
  }
  // sync confidence: fraction of sync bits matched after polarity resolution
  let syncMatch = 0;
  for (let k = 0; k < SYNC_BITS.length; k++) {
    const bit = polarity * syncSoft[k]! > 0 ? 1 : 0;
    if (bit === SYNC_BITS[k]) syncMatch++;
  }
  const syncScore = syncMatch / SYNC_BITS.length;
  if (syncScore < 0.75) {
    return {
      ok: false, payload: null, blockLen: 0, syncOffset: bestOff,
      syncScore, rsErrors: 0, cfoHz, reason: `weak sync ${syncScore.toFixed(2)}`,
    };
  }

  // ---- read the triplicated header -> blockLen ----
  const hdrStart = dataStart + SYNC_BITS.length * sps;
  const hdrBits = new Uint8Array(HEADER_BITS * HEADER_REPEAT);
  for (let k = 0; k < hdrBits.length; k++) {
    const sv = symbolCorrelate(workRe, workIm, hdrStart + k * sps, sps, o.baseFreqHz, o.iqRateHz);
    hdrBits[k] = sv.re * refRe + sv.im * refIm > 0 ? 1 : 0;
  }
  const lens: number[] = [];
  for (let r = 0; r < HEADER_REPEAT; r++) {
    let v = 0;
    for (let b = 0; b < HEADER_BITS; b++) v = (v << 1) | hdrBits[r * HEADER_BITS + b]!;
    lens.push(v);
  }
  const blockLen = majority(lens);
  if (blockLen <= 2 || blockLen > 255) {
    return {
      ok: false, payload: null, blockLen, syncOffset: bestOff,
      syncScore, rsErrors: 0, cfoHz, reason: `bad header len ${blockLen}`,
    };
  }

  // ---- read the body coded bits -> FEC -> payload + crc ----
  const codedBits = codedBitLength(blockLen, o.fec);
  const bodyStart = hdrStart + hdrBits.length * sps;
  if (bodyStart + codedBits * sps > workRe.length) {
    return {
      ok: false, payload: null, blockLen, syncOffset: bestOff,
      syncScore, rsErrors: 0, cfoHz, reason: 'frame truncated',
    };
  }
  const codedStream = new Uint8Array(codedBits);
  for (let k = 0; k < codedBits; k++) {
    const sv = symbolCorrelate(workRe, workIm, bodyStart + k * sps, sps, o.baseFreqHz, o.iqRateHz);
    codedStream[k] = sv.re * refRe + sv.im * refIm > 0 ? 1 : 0;
  }
  let body: Uint8Array;
  let rsErrors = 0;
  try {
    const dec = decodeFec(codedStream, o.fec, blockLen);
    body = dec.block;
    rsErrors = dec.rsErrors;
  } catch (e) {
    return {
      ok: false, payload: null, blockLen, syncOffset: bestOff,
      syncScore, rsErrors: 0, cfoHz, reason: `fec: ${(e as Error).message}`,
    };
  }
  const payload = body.subarray(0, body.length - 2);
  const gotCrc = ((body[body.length - 2]! << 8) | body[body.length - 1]!) & 0xffff;
  const wantCrc = crc16Bytes(payload);
  if (gotCrc !== wantCrc) {
    return {
      ok: false, payload: null, blockLen, syncOffset: bestOff,
      syncScore, rsErrors, cfoHz, reason: 'crc mismatch',
    };
  }
  return {
    ok: true, payload: payload.slice(), blockLen, syncOffset: bestOff,
    syncScore, rsErrors, cfoHz,
  };
}

function emptyFail(syncScore: number, reason: string): BpskDemodResult {
  return {
    ok: false, payload: null, blockLen: 0, syncOffset: -1,
    syncScore, rsErrors: 0, cfoHz: 0, reason,
  };
}

function majority(vals: number[]): number {
  const counts = new Map<number, number>();
  let best = vals[0]!;
  let bestC = 0;
  for (const v of vals) {
    const c = (counts.get(v) ?? 0) + 1;
    counts.set(v, c);
    if (c > bestC) {
      bestC = c;
      best = v;
    }
  }
  return best;
}
