// @a0n/.ignored_gnosis/rtlsdr-mock-sim/fec-codec.ts
/**
 * Concatenated FEC orchestrator: payload -> RS(outer) -> block interleave ->
 * convolutional(inner) -> coded bits, and the exact inverse. The single seam the
 * modem calls. Lengths are deterministic from the payload length + profile, so
 * the decoder needs only the original block length (carried in the modem frame).
 */
import {
  bitsToBytes,
  bytesToBits,
  convEncode,
  K,
  viterbiDecode,
} from './fec-convolutional';
import { blockInterleave, blockDeinterleave } from './interleaver';
import { rsDecode, rsEncode } from './fec-reed-solomon';

const MEM = K - 1;

export interface FecProfile {
  readonly name: string;
  readonly rsParity: number; // 0 = no RS
  readonly conv: boolean;
  readonly interleaveCols: number; // 1 = no interleave
}

export const FEC_PROFILES = {
  none: { name: 'none', rsParity: 0, conv: false, interleaveCols: 1 },
  conv: { name: 'conv', rsParity: 0, conv: true, interleaveCols: 1 },
  rs: { name: 'rs', rsParity: 20, conv: false, interleaveCols: 8 },
  'rs+conv': { name: 'rs+conv', rsParity: 20, conv: true, interleaveCols: 8 },
} as const satisfies Record<string, FecProfile>;

export type FecProfileName = keyof typeof FEC_PROFILES;

/**
 * Resolves the Fec Profile.
 */
export function resolveFecProfile(name: string): FecProfile {
  const p = (FEC_PROFILES as Record<string, FecProfile>)[name];
  if (!p) throw new Error(`unknown fec profile ${name}`);
  return p;
}

interface Dims {
  readonly postRsLen: number;
  readonly rows: number;
  readonly cols: number;
  readonly interleavedLen: number;
  readonly preConvBits: number;
}

function dims(blockLen: number, p: FecProfile): Dims {
  const postRsLen = blockLen + p.rsParity;
  const cols = p.interleaveCols > 1 ? p.interleaveCols : 1;
  const rows = cols > 1 ? Math.ceil(postRsLen / cols) : postRsLen;
  const interleavedLen = cols > 1 ? rows * cols : postRsLen;
  return { postRsLen, rows, cols, interleavedLen, preConvBits: interleavedLen * 8 };
}

/** Encode a payload block to coded bits ({0,1}). */
export function encodeFec(block: Uint8Array, p: FecProfile): Uint8Array {
  let bytes: Uint8Array = block;
  if (p.rsParity > 0) bytes = rsEncode(bytes, p.rsParity);
  const d = dims(block.length, p);
  if (p.interleaveCols > 1) bytes = blockInterleave(bytes, d.rows, d.cols);
  let bits = bytesToBits(bytes);
  if (p.conv) bits = convEncode(bits);
  return bits;
}

/** Decode coded bits back to the payload block. Throws if uncorrectable. */
export function decodeFec(
  bits: Uint8Array,
  p: FecProfile,
  blockLen: number,
): { block: Uint8Array; rsErrors: number } {
  const d = dims(blockLen, p);
  let interBits: Uint8Array;
  if (p.conv) {
    const expected = (d.preConvBits + MEM) * 2;
    if (bits.length !== expected) {
      throw new Error(`decodeFec: expected ${expected} coded bits, got ${bits.length}`);
    }
    interBits = viterbiDecode(bits, d.preConvBits);
  } else {
    interBits = bits.subarray(0, d.preConvBits);
  }
  let bytes = bitsToBytes(interBits).subarray(0, d.interleavedLen);
  if (p.interleaveCols > 1) {
    bytes = blockDeinterleave(bytes, d.rows, d.cols, d.postRsLen);
  } else {
    bytes = bytes.subarray(0, d.postRsLen);
  }
  if (p.rsParity > 0) {
    const out = rsDecode(bytes, p.rsParity);
    return { block: out.message.subarray(0, blockLen), rsErrors: out.errorsCorrected };
  }
  return { block: bytes.subarray(0, blockLen), rsErrors: 0 };
}

/** Coded-bit count for a payload length under a profile (modem sizing). */
export function codedBitLength(blockLen: number, p: FecProfile): number {
  const d = dims(blockLen, p);
  return p.conv ? (d.preConvBits + MEM) * 2 : d.preConvBits;
}
