// @a0n/.ignored_gnosis/rtlsdr-mock-sim/fec-convolutional.ts
/**
 * Rate-1/2, constraint-length-7 convolutional code with hard-decision Viterbi.
 * Generator polynomials G0=0o171, G1=0o133 -- the CCSDS / Voyager standard, the
 * cheapest large coding-gain win (~5 dB on AWGN). Zero-terminated (6 flush bits).
 *
 * Bits are represented as Uint8Array of {0,1}. Two output bits per input bit.
 */

export const K = 7;
const MEM = K - 1; // 6
const NSTATES = 1 << MEM; // 64
export const G0 = 0o171;
export const G1 = 0o133;

function parity7(v: number): number {
  let x = v & 0x7f;
  x ^= x >> 4;
  x ^= x >> 2;
  x ^= x >> 1;
  return x & 1;
}

// Precompute transition outputs: out[state][bit] = [o0,o1], next[state][bit].
const OUT0 = new Uint8Array(NSTATES * 2);
const OUT1 = new Uint8Array(NSTATES * 2);
const NEXT = new Uint8Array(NSTATES * 2);
for (let s = 0; s < NSTATES; s++) {
  for (let b = 0; b < 2; b++) {
    const v = (s << 1) | b; // 7-bit window
    OUT0[s * 2 + b] = parity7(v & G0);
    OUT1[s * 2 + b] = parity7(v & G1);
    NEXT[s * 2 + b] = v & (NSTATES - 1);
  }
}

/** Encode {0,1} bits -> {0,1} bits (length 2*(n+MEM) after zero flush). */
export function convEncode(bits: Uint8Array): Uint8Array {
  const n = bits.length;
  const out = new Uint8Array((n + MEM) * 2);
  let state = 0;
  let o = 0;
  const step = (b: number): void => {
    out[o++] = OUT0[state * 2 + b]!;
    out[o++] = OUT1[state * 2 + b]!;
    state = NEXT[state * 2 + b]!;
  };
  for (let i = 0; i < n; i++) step(bits[i]! & 1);
  for (let i = 0; i < MEM; i++) step(0); // flush to state 0
  return out;
}

/** Hard-decision Viterbi decode. Input length must be 2*(msgBits+MEM). */
export function viterbiDecode(coded: Uint8Array, msgBits: number): Uint8Array {
  const steps = msgBits + MEM;
  if (coded.length !== steps * 2) {
    throw new Error(`viterbiDecode: expected ${steps * 2} coded bits, got ${coded.length}`);
  }
  const INF = 1e9;
  let cost = new Float64Array(NSTATES).fill(INF);
  cost[0] = 0;
  // traceback: for each step, store predecessor bit per state
  const tbBit = new Uint8Array(steps * NSTATES);
  const tbPrev = new Uint16Array(steps * NSTATES);
  for (let t = 0; t < steps; t++) {
    const r0 = coded[t * 2]! & 1;
    const r1 = coded[t * 2 + 1]! & 1;
    const next = new Float64Array(NSTATES).fill(INF);
    const base = t * NSTATES;
    for (let s = 0; s < NSTATES; s++) {
      const c = cost[s]!;
      if (c >= INF) continue;
      for (let b = 0; b < 2; b++) {
        const ns = NEXT[s * 2 + b]!;
        const m = (OUT0[s * 2 + b]! ^ r0) + (OUT1[s * 2 + b]! ^ r1);
        const nc = c + m;
        if (nc < next[ns]!) {
          next[ns] = nc;
          tbBit[base + ns] = b;
          tbPrev[base + ns] = s;
        }
      }
    }
    cost = next;
  }
  // terminate at state 0 (zero-flushed)
  const bits = new Uint8Array(msgBits);
  let state = 0;
  for (let t = steps - 1; t >= 0; t--) {
    const base = t * NSTATES;
    const b = tbBit[base + state]!;
    const prev = tbPrev[base + state]!;
    if (t < msgBits) bits[t] = b;
    state = prev;
  }
  return bits;
}

/** Pack {0,1} bits (MSB-first per byte) into bytes. */
export function bitsToBytes(bits: Uint8Array): Uint8Array {
  const out = new Uint8Array(Math.ceil(bits.length / 8));
  for (let i = 0; i < bits.length; i++) {
    if (bits[i]! & 1) out[i >> 3]! |= 0x80 >> (i & 7);
  }
  return out;
}

/** Unpack bytes into {0,1} bits (MSB-first), length = 8*bytes. */
export function bytesToBits(bytes: Uint8Array): Uint8Array {
  const out = new Uint8Array(bytes.length * 8);
  for (let i = 0; i < bytes.length; i++) {
    const v = bytes[i]!;
    for (let b = 0; b < 8; b++) out[i * 8 + b] = (v >> (7 - b)) & 1;
  }
  return out;
}
