// @a0n/.ignored_gnosis/rtlsdr-mock-sim/fec-reed-solomon.ts
/**
 * Systematic Reed-Solomon over GF(2^8), primitive poly 0x11d, c=0 syndromes.
 * Outer code in the CCSDS concatenation: corrects up to floor(nsym/2) byte
 * errors, mopping up the bursts the inner Viterbi decoder leaves behind.
 *
 * Faithful port of the standard "Reed-Solomon for coders" algorithm
 * (exp/log GF tables, Berlekamp-Massey locator, Chien search, Forney magnitudes).
 * Polynomials are number[] with index 0 = highest degree.
 */

const PRIM = 0x11d;
const EXP = new Uint8Array(512);
const LOG = new Uint8Array(256);
(() => {
  let x = 1;
  for (let i = 0; i < 255; i++) {
    EXP[i] = x;
    LOG[x] = i;
    x <<= 1;
    if (x & 0x100) x ^= PRIM;
  }
  for (let i = 255; i < 512; i++) EXP[i] = EXP[i - 255]!;
})();

const gfMul = (a: number, b: number): number =>
  a === 0 || b === 0 ? 0 : EXP[LOG[a]! + LOG[b]!]!;
const gfDiv = (a: number, b: number): number => {
  if (b === 0) throw new Error('gfDiv by zero');
  if (a === 0) return 0;
  return EXP[(LOG[a]! + 255 - LOG[b]!) % 255]!;
};
const gfPow = (a: number, p: number): number => {
  const e = (((LOG[a]! * p) % 255) + 255) % 255;
  return EXP[e]!;
};
const gfInv = (a: number): number => EXP[255 - LOG[a]!]!;

function polyScale(p: number[], x: number): number[] {
  return p.map((c) => gfMul(c, x));
}
function polyAdd(p: number[], q: number[]): number[] {
  const r = new Array<number>(Math.max(p.length, q.length)).fill(0);
  for (let i = 0; i < p.length; i++) r[i + r.length - p.length] = p[i]!;
  for (let i = 0; i < q.length; i++) r[i + r.length - q.length]! ^= q[i]!;
  return r;
}
function polyMul(p: number[], q: number[]): number[] {
  const r = new Array<number>(p.length + q.length - 1).fill(0);
  for (let j = 0; j < q.length; j++) {
    for (let i = 0; i < p.length; i++) {
      r[i + j]! ^= gfMul(p[i]!, q[j]!);
    }
  }
  return r;
}
function polyEval(p: number[], x: number): number {
  let y = p[0]!;
  for (let i = 1; i < p.length; i++) y = gfMul(y, x) ^ p[i]!;
  return y;
}
function polyDiv(dividend: number[], divisor: number[]): { rem: number[] } {
  const out = dividend.slice();
  for (let i = 0; i < dividend.length - (divisor.length - 1); i++) {
    const coef = out[i]!;
    if (coef !== 0) {
      for (let j = 1; j < divisor.length; j++) {
        if (divisor[j] !== 0) out[i + j]! ^= gfMul(divisor[j]!, coef);
      }
    }
  }
  const sep = out.length - (divisor.length - 1);
  return { rem: out.slice(sep) };
}

function generatorPoly(nsym: number): number[] {
  let g = [1];
  for (let i = 0; i < nsym; i++) g = polyMul(g, [1, gfPow(2, i)]);
  return g;
}

/** Systematic encode: returns data followed by nsym parity bytes. */
export function rsEncode(data: Uint8Array, nsym: number): Uint8Array {
  if (data.length + nsym > 255) throw new Error('rsEncode: codeword > 255');
  const gen = generatorPoly(nsym);
  const out = new Array<number>(data.length + nsym).fill(0);
  for (let i = 0; i < data.length; i++) out[i] = data[i]!;
  for (let i = 0; i < data.length; i++) {
    const coef = out[i]!;
    if (coef !== 0) {
      for (let j = 1; j < gen.length; j++) out[i + j]! ^= gfMul(gen[j]!, coef);
    }
  }
  for (let i = 0; i < data.length; i++) out[i] = data[i]!; // restore message bytes
  return Uint8Array.from(out);
}

function calcSyndromes(msg: number[], nsym: number): number[] {
  const synd = [0];
  for (let i = 0; i < nsym; i++) synd.push(polyEval(msg, gfPow(2, i)));
  return synd;
}

function findErrorLocator(synd: number[], nsym: number): number[] {
  let errLoc = [1];
  let oldLoc = [1];
  const syndShift = synd.length > nsym ? synd.length - nsym : 0;
  for (let i = 0; i < nsym; i++) {
    const K = i + syndShift;
    let delta = synd[K]!;
    for (let j = 1; j < errLoc.length; j++) {
      delta ^= gfMul(errLoc[errLoc.length - (j + 1)]!, synd[K - j]!);
    }
    oldLoc = oldLoc.concat([0]);
    if (delta !== 0) {
      if (oldLoc.length > errLoc.length) {
        const newLoc = polyScale(oldLoc, delta);
        oldLoc = polyScale(errLoc, gfInv(delta));
        errLoc = newLoc;
      }
      errLoc = polyAdd(errLoc, polyScale(oldLoc, delta));
    }
  }
  while (errLoc.length && errLoc[0] === 0) errLoc.shift();
  const errs = errLoc.length - 1;
  if (errs * 2 > nsym) throw new Error('rsDecode: too many errors');
  return errLoc;
}

function findErrors(errLocRev: number[], nmess: number): number[] {
  const errs = errLocRev.length - 1;
  const errPos: number[] = [];
  for (let i = 0; i < nmess; i++) {
    if (polyEval(errLocRev, gfPow(2, i)) === 0) errPos.push(nmess - 1 - i);
  }
  if (errPos.length !== errs) throw new Error('rsDecode: could not locate errors');
  return errPos;
}

function findErrataLocator(coefPos: number[]): number[] {
  let eLoc = [1];
  for (const p of coefPos) eLoc = polyMul(eLoc, polyAdd([1], [gfPow(2, p), 0]));
  return eLoc;
}

function findErrorEvaluator(synd: number[], errLoc: number[], nsym: number): number[] {
  const divisor = [1, ...new Array<number>(nsym + 1).fill(0)];
  return polyDiv(polyMul(synd, errLoc), divisor).rem;
}

function correctErrata(msg: number[], synd: number[], errPos: number[]): number[] {
  const coefPos = errPos.map((p) => msg.length - 1 - p);
  const errLoc = findErrataLocator(coefPos);
  const errEval = findErrorEvaluator(synd.slice().reverse(), errLoc, errLoc.length - 1).reverse();
  const X: number[] = [];
  for (let i = 0; i < coefPos.length; i++) {
    const l = 255 - coefPos[i]!;
    X.push(gfPow(2, -l));
  }
  const E = new Array<number>(msg.length).fill(0);
  for (let i = 0; i < X.length; i++) {
    const Xi = X[i]!;
    const XiInv = gfInv(Xi);
    let errLocPrime = 1;
    for (let j = 0; j < X.length; j++) {
      if (j !== i) errLocPrime = gfMul(errLocPrime, 1 ^ gfMul(XiInv, X[j]!));
    }
    if (errLocPrime === 0) throw new Error('rsDecode: zero error locator derivative');
    let y = polyEval(errEval.slice().reverse(), XiInv);
    y = gfMul(gfPow(Xi, 1), y);
    E[errPos[i]!] = gfDiv(y, errLocPrime);
  }
  return polyAdd(msg, E);
}

export interface RsDecodeResult {
  readonly message: Uint8Array;
  readonly errorsCorrected: number;
}

/** Decode a (data+parity) codeword. Throws if uncorrectable. */
export function rsDecode(codeword: Uint8Array, nsym: number): RsDecodeResult {
  if (codeword.length > 255) throw new Error('rsDecode: codeword > 255');
  let msg = Array.from(codeword);
  let synd = calcSyndromes(msg, nsym);
  if (Math.max(...synd) === 0) {
    return { message: Uint8Array.from(msg.slice(0, -nsym)), errorsCorrected: 0 };
  }
  const errLoc = findErrorLocator(synd, nsym);
  const errPos = findErrors(errLoc.slice().reverse(), msg.length);
  msg = correctErrata(msg, synd, errPos);
  synd = calcSyndromes(msg, nsym);
  if (Math.max(...synd) > 0) throw new Error('rsDecode: could not correct message');
  return { message: Uint8Array.from(msg.slice(0, -nsym)), errorsCorrected: errPos.length };
}

/** Max correctable byte errors for nsym parity bytes. */
export const rsMaxErrors = (nsym: number): number => Math.floor(nsym / 2);
