// @a0n/.ignored_gnosis/rtlsdr-mock-sim/fm-demod.ts
import { complexFromU8Pair } from './iq';

export interface DemodToWavPcm16Options {
  readonly iqSampleRateHz: number;
  readonly audioSampleRateHz: number;
}

/** Quadrature FM discriminator on rtl_u8 IQ, box-averaged to audio rate. */
export function demodulateFmDiscriminatorDecimated(
  iqU8: Uint8Array,
  opts: DemodToWavPcm16Options,
): Float32Array {
  if (iqU8.length < 4 || iqU8.length % 2 !== 0) {
    throw new Error('demodulateFmDiscriminatorDecimated: need even u8 length >= 4');
  }
  const ratio = Math.max(2, Math.round(opts.iqSampleRateHz / opts.audioSampleRateHz));
  const nPairs = iqU8.length / 2;
  const disc: number[] = [];
  let prev = complexFromU8Pair(iqU8[0]!, iqU8[1]!);
  for (let p = 2; p + 1 < iqU8.length; p += 2) {
    const z = complexFromU8Pair(iqU8[p]!, iqU8[p + 1]!);
    const pr = prev.re * z.re + prev.im * z.im;
    const pi = prev.im * z.re - prev.re * z.im;
    disc.push(Math.atan2(pi, pr));
    prev = z;
  }
  const outLen = Math.floor(disc.length / ratio);
  const out = new Float32Array(outLen);
  for (let j = 0; j < outLen; j++) {
    let s = 0;
    for (let k = 0; k < ratio; k++) {
      s += disc[j * ratio + k]!;
    }
    out[j] = s / ratio;
  }
  return out;
}

/**
 * Handles the gnosis high Pass One Pole workflow.
 */
export function highPassOnePole(x: Float32Array, alpha: number): Float32Array {
  const y = new Float32Array(x.length);
  let xp = 0;
  let yp = 0;
  for (let i = 0; i < x.length; i++) {
    const xi = x[i]!;
    y[i] = alpha * (yp + xi - xp);
    xp = xi;
    yp = y[i]!;
  }
  return y;
}

/**
 * Handles the gnosis rms workflow.
 */
export function rms(x: Float32Array): number {
  if (x.length === 0) return 0;
  let s = 0;
  for (let i = 0; i < x.length; i++) {
    const v = x[i]!;
    s += v * v;
  }
  return Math.sqrt(s / x.length);
}

/**
 * Handles the gnosis pcm16 Mono From Float Audio workflow.
 */
export function pcm16MonoFromFloatAudio(
  x: Float32Array,
  gain: number,
): Int16Array {
  const y = new Int16Array(x.length);
  for (let i = 0; i < x.length; i++) {
    const v = Math.max(-1, Math.min(1, x[i]! * gain));
    y[i] = Math.round(v * 32767);
  }
  return y;
}

/**
 * Handles the gnosis score Voice Band Energy workflow.
 */
export function scoreVoiceBandEnergy(x: Float32Array): number {
  const hp = highPassOnePole(x, 0.97);
  return rms(hp);
}

/**
 * Mean IQ discriminator `arg(b·conj(a))` (rad/sample) × fs / (2π) ≈ dominant spin (Hz).
 * Accurate for **CW / slow-spin** baseband; biased under wide FM — use only as a lab hint there.
 */
export function estimateMeanInstFreqHz(
  iqU8: Uint8Array,
  sampleRateHz: number,
  maxPairs: number,
): number {
  const n = Math.min(maxPairs, Math.floor(iqU8.length / 2) - 1);
  if (n < 8) return 0;
  let s = 0;
  for (let i = 0; i < n; i++) {
    const a = complexFromU8Pair(iqU8[2 * i]!, iqU8[2 * i + 1]!);
    const b = complexFromU8Pair(iqU8[2 * i + 2]!, iqU8[2 * i + 3]!);
    const pr = b.re * a.re + b.im * a.im;
    const pi = b.im * a.re - b.re * a.im;
    s += Math.atan2(pi, pr);
  }
  return (s / n) * (sampleRateHz / (Math.PI * 2));
}

/** Coherent score for a known message tone (Hz) in demodulated audio. */
export function scoreToneProjection(
  x: Float32Array,
  toneHz: number,
  sampleRateHz: number,
): number {
  if (x.length === 0) return 0;
  let c = 0;
  let s = 0;
  const w = (Math.PI * 2 * toneHz) / sampleRateHz;
  for (let i = 0; i < x.length; i++) {
    const xi = x[i]!;
    c += xi * Math.cos(w * i);
    s += xi * Math.sin(w * i);
  }
  return (c * c + s * s) / x.length;
}
