// @a0n/.ignored_gnosis/rtlsdr-mock-sim/link-budget.ts
/**
 * Deep-space RF link budget (Friis) for the Mars-channel cache-key teleport.
 *
 * Real physics, standard formulas. The "survives Mars" claim made elsewhere is
 * an Eb/N0-EQUIVALENCE claim: this module computes the Eb/N0 a given distance +
 * dish + power + bit-rate would actually deliver, the bench then reproduces that
 * Eb/N0 by attenuating the Pluto, and the FEC must close the gap. If the margin
 * does not close we report marginClosesDb < 0 plus the dish/power that would --
 * we never silently lower fidelity.
 *
 * Required-Eb/N0 numbers are standard AWGN approximations at BER 1e-6 and are
 * documented inline; they are conservative reference values, not measured.
 */

export const BOLTZMANN_DBW_PER_HZ_K = -228.6; // 10*log10(k), k=1.380649e-23 J/K
export const SPEED_OF_LIGHT_M_S = 299_792_458;

export type Modulation = 'mfsk8' | 'mfsk4' | 'bpsk';
export type FecProfile = 'none' | 'conv' | 'rs' | 'rs+conv';

/** Named RF bands used by the demo. ism433 is the bench band; X-band is real DSN. */
export const BANDS = {
  'ism433': 433_920_000,
  'x-up': 7_145_000_000,
  'x-down': 8_420_000_000,
  'ka-down': 32_000_000_000,
} as const;
export type BandName = keyof typeof BANDS;

/** Free-space path loss (dB). d in meters, f in Hz. FSPL = 20log10(4*pi*d*f/c). */
export function friisPathLossDb(distanceM: number, freqHz: number): number {
  if (!(distanceM > 0) || !(freqHz > 0)) {
    throw new Error(`friisPathLossDb: bad d=${distanceM} f=${freqHz}`);
  }
  return (
    20 * Math.log10(distanceM) +
    20 * Math.log10(freqHz) +
    20 * Math.log10((4 * Math.PI) / SPEED_OF_LIGHT_M_S)
  );
}

/** Parabolic dish gain (dBi). G = eff * (pi*D*f/c)^2. */
export function dishGainDbi(diameterM: number, freqHz: number, efficiency = 0.55): number {
  if (!(diameterM > 0) || !(freqHz > 0)) {
    throw new Error(`dishGainDbi: bad D=${diameterM} f=${freqHz}`);
  }
  const lambda = SPEED_OF_LIGHT_M_S / freqHz;
  const gain = efficiency * Math.pow((Math.PI * diameterM) / lambda, 2);
  return 10 * Math.log10(gain);
}

/** Noise power spectral density N0 (dBW/Hz) for a system noise temperature (K). */
export function noisePsdDbwPerHz(systemNoiseTempK: number): number {
  if (!(systemNoiseTempK > 0)) {
    throw new Error(`noisePsdDbwPerHz: bad T=${systemNoiseTempK}`);
  }
  return BOLTZMANN_DBW_PER_HZ_K + 10 * Math.log10(systemNoiseTempK);
}

/**
 * Uncoded required Eb/N0 (dB) for BER ~1e-6, AWGN. Standard reference values:
 * BPSK ~10.5; non-coherent 4-FSK ~10.8; non-coherent 8-FSK ~9.3 (orthogonal
 * M-FSK improves with M at fixed BER, the classic power-vs-bandwidth trade).
 */
export function uncodedRequiredEbN0Db(mod: Modulation): number {
  switch (mod) {
    case 'bpsk':
      return 10.5;
    case 'mfsk4':
      return 10.8;
    case 'mfsk8':
      return 9.3;
  }
}

/**
 * Coding gain (dB) at BER ~1e-6, AWGN. Standard reference values:
 * conv K=7 r=1/2 + Viterbi ~5.2; RS(255,223) ~2.5 alone over a clean channel but
 * its real value is burst cleanup; concatenated RS+conv (the CCSDS/Voyager stack)
 * ~7.5 total. Whip fork/race/fold diversity is modeled separately as combining gain.
 */
export function codingGainDb(fec: FecProfile): number {
  switch (fec) {
    case 'none':
      return 0;
    case 'conv':
      return 5.2;
    case 'rs':
      return 2.5;
    case 'rs+conv':
      return 7.5;
  }
}

/** Soft-combine diversity gain (dB) for N Whip replicas. Capped, ~10log10(N). */
export function whipDiversityGainDb(replicas: number): number {
  if (!(replicas >= 1)) return 0;
  return 10 * Math.log10(replicas);
}

/**
 * Handles the gnosis bits Per Symbol workflow.
 */
export function bitsPerSymbol(mod: Modulation): number {
  switch (mod) {
    case 'bpsk':
      return 1;
    case 'mfsk4':
      return 2;
    case 'mfsk8':
      return 3;
  }
}

export interface LinkBudgetParams {
  readonly distanceM: number;
  readonly freqHz: number;
  readonly txPowerDbm: number;
  readonly txDishM: number;
  readonly rxDishM: number;
  readonly dishEfficiency?: number;
  readonly systemNoiseTempK: number;
  readonly dataRateBps: number;
  readonly mod: Modulation;
  readonly fec: FecProfile;
  readonly whipReplicas?: number;
  readonly targetBer?: number;
}

export interface LinkBudgetResult {
  readonly fsplDb: number;
  readonly txGainDbi: number;
  readonly rxGainDbi: number;
  readonly eirpDbm: number;
  readonly rxPowerDbm: number;
  readonly noisePsdDbmPerHz: number;
  readonly receivedEbN0Db: number;
  readonly requiredEbN0Db: number;
  readonly codingGainDb: number;
  readonly whipGainDb: number;
  readonly marginDb: number;
  readonly marginCloses: boolean;
}

/** Full one-way link budget. dBm path (dBW + 30). */
export function linkBudget(p: LinkBudgetParams): LinkBudgetResult {
  const eff = p.dishEfficiency ?? 0.55;
  const fsplDb = friisPathLossDb(p.distanceM, p.freqHz);
  const txGainDbi = dishGainDbi(p.txDishM, p.freqHz, eff);
  const rxGainDbi = dishGainDbi(p.rxDishM, p.freqHz, eff);
  const eirpDbm = p.txPowerDbm + txGainDbi;
  const rxPowerDbm = eirpDbm - fsplDb + rxGainDbi;
  const n0DbmPerHz = noisePsdDbwPerHz(p.systemNoiseTempK) + 30; // dBW/Hz -> dBm/Hz
  // Eb/N0 = Pr - 10log10(Rb) - N0
  const receivedEbN0Db = rxPowerDbm - 10 * Math.log10(p.dataRateBps) - n0DbmPerHz;
  const gain = codingGainDb(p.fec);
  const whip = whipDiversityGainDb(p.whipReplicas ?? 1);
  const requiredEbN0Db = uncodedRequiredEbN0Db(p.mod) - gain - whip;
  const marginDb = receivedEbN0Db - requiredEbN0Db;
  return {
    fsplDb,
    txGainDbi,
    rxGainDbi,
    eirpDbm,
    rxPowerDbm,
    noisePsdDbmPerHz: n0DbmPerHz,
    receivedEbN0Db,
    requiredEbN0Db,
    codingGainDb: gain,
    whipGainDb: whip,
    marginDb,
    marginCloses: marginDb >= 0,
  };
}

/**
 * Per-component AWGN RMS that fills the SDR ADC without clipping. Real SDR AGC
 * sets the noise floor to fill the converter and the wanted signal sits well
 * below full scale, so we fix the noise here and scale the signal for Eb/N0.
 */
export const ADC_NOISE_FILL_RMS = 0.25;

/**
 * Bench Eb/N0-equivalence: with the noise fixed at `noiseFillRms`, return the
 * modulated-signal amplitude that yields a target received Eb/N0 after the
 * per-symbol matched filter (a single-bin DFT over samplesPerSymbol).
 *
 * Matched-filter output SNR = sps * A^2 / (2 * sigma^2) = Es/N0, so
 *   A = sigma * sqrt(2 * EsN0 / sps),  EsN0 = EbN0 * bitsPerSymbol.
 * This keeps signal + noise inside the u8 [-1,1] range so the bench is real AWGN,
 * not clipping. The same target maps to a Pluto hardwaregain on real silicon.
 */
export function signalAmplitudeForEbN0(
  ebN0Db: number,
  mod: Modulation,
  samplesPerSymbol: number,
  noiseFillRms: number = ADC_NOISE_FILL_RMS,
): number {
  const esN0 = Math.pow(10, (ebN0Db + 10 * Math.log10(bitsPerSymbol(mod))) / 10);
  return noiseFillRms * Math.sqrt((2 * esN0) / samplesPerSymbol);
}

export const LINK_BUDGET_CERT_VERSION = 'gnosis.mars-link-budget.v1';

export interface BenchEmulation {
  readonly method: 'pluto-hardwaregain' | 'software-awgn';
  readonly targetEbN0Db: number;
  readonly noiseFillRms: number;
  readonly signalAmplitude: number;
  readonly note: string;
}

export interface LinkBudgetCertificate {
  readonly certificateVersion: typeof LINK_BUDGET_CERT_VERSION;
  readonly label: string;
  readonly params: LinkBudgetParams;
  readonly result: LinkBudgetResult;
  readonly benchEmulation: BenchEmulation;
  readonly honest: string;
}

/**
 * Builds the Link Budget Certificate.
 */
export function buildLinkBudgetCertificate(
  label: string,
  params: LinkBudgetParams,
  samplesPerSymbol: number,
): LinkBudgetCertificate {
  const result = linkBudget(params);
  return {
    certificateVersion: LINK_BUDGET_CERT_VERSION,
    label,
    params,
    result,
    benchEmulation: {
      method: 'software-awgn',
      targetEbN0Db: result.receivedEbN0Db,
      noiseFillRms: ADC_NOISE_FILL_RMS,
      signalAmplitude: signalAmplitudeForEbN0(
        result.receivedEbN0Db,
        params.mod,
        samplesPerSymbol,
      ),
      note:
        'Bench reproduces the link-budget receivedEbN0Db by transmitting at signalAmplitude ' +
        'against AWGN fixed at noiseFillRms (software), or the equivalent Pluto hardwaregain. ' +
        'Physics is real; distance is emulated.',
    },
    honest:
      'Eb/N0-equivalence: the bench is not 3-22 light-minutes of vacuum. The link ' +
      'budget is real; the survival claim is that FEC closes the same Eb/N0 the ' +
      'modeled distance/dish/power would deliver.',
  };
}
