// @a0n/.ignored_gnosis/rtlsdr-mock-sim/mars-light-time.ts
/**
 * One-way light-time model for the interplanetary cache-key teleport.
 *
 * The teleport path never pays this delay; it is applied only to the COLD
 * round-trip A/B baseline (see mars-cold-roundtrip.ts) so the certificate can
 * record the true modeled latency a real link would suffer.
 *
 * Distances are Earth<->body ranges. For planets we expose min/mean/max regimes
 * because the range varies hugely over a synodic period (Mars: ~3 to ~22 min).
 */

/** Speed of light in km/s (exact, by SI definition / 1000). */
export const SPEED_OF_LIGHT_KM_S = 299_792.458;

export type DistanceRegime = 'opposition' | 'mean' | 'conjunction';

/** Earth<->Mars range (km) per regime. opposition = closest, conjunction = farthest. */
export const MARS_RANGE_KM: Readonly<Record<DistanceRegime, number>> = {
  opposition: 54_600_000,
  mean: 225_000_000,
  conjunction: 401_000_000,
};

/** One-way light-time (seconds) for a range in km. */
export function lightTimeSecFromKm(rangeKm: number): number {
  if (!(rangeKm >= 0) || !Number.isFinite(rangeKm)) {
    throw new Error(`lightTimeSecFromKm: bad rangeKm ${rangeKm}`);
  }
  return rangeKm / SPEED_OF_LIGHT_KM_S;
}

/** Round-trip light-time (seconds) for a range in km. */
export function roundTripSecFromKm(rangeKm: number): number {
  return 2 * lightTimeSecFromKm(rangeKm);
}

/** Mars one-way light-time (seconds) for a named regime. */
export function marsDelayFromRegime(regime: DistanceRegime): number {
  return lightTimeSecFromKm(MARS_RANGE_KM[regime]);
}

export interface LightTimeReport {
  readonly rangeKm: number;
  readonly oneWaySec: number;
  readonly roundTripSec: number;
  readonly oneWayMinutes: number;
}

/**
 * Handles the gnosis light Time Report workflow.
 */
export function lightTimeReport(rangeKm: number): LightTimeReport {
  const oneWaySec = lightTimeSecFromKm(rangeKm);
  return {
    rangeKm,
    oneWaySec,
    roundTripSec: 2 * oneWaySec,
    oneWayMinutes: oneWaySec / 60,
  };
}
