// @a0n/.ignored_gnosis/rtlsdr-mock-sim/pluto-tx.ts
/**
 * PlutoSDR transmit helpers for the Mars-channel loopback. Converts offset-binary
 * u8 IQ (the modem wire format) to the signed-int16 interleaved buffer libiio's
 * iio_writedev streams to cf-ad9361-dds-core-lpc, and builds the iio_attr / iio_writedev
 * commands. libiio here is a from-source build (macOS framework) under
 * /tmp/libiio_build -- ephemeral, so we fail closed with a clear message if it is gone.
 */
import { existsSync, writeFileSync } from 'node:fs';
import { complexFromU8Pair } from './iq';

export const LIBIIO_BUILD_DIR = '/tmp/libiio_build/build';
export const LIBIIO_FRAMEWORK = `${LIBIIO_BUILD_DIR}/iio.framework/iio`;
/** Last-resort default; the Pluto's USB address varies per enumeration. */
export const PLUTO_URI = 'usb:0.5.5';

/**
 * Resolve the libiio URI to target. Honors GNOSIS_PLUTO_URI (set by
 * setup-sdr-toolchain.sh / the operator after auto-discovery); falls back to the
 * static default only when unset.
 */
export function resolvePlutoUri(): string {
  const env = (process.env.GNOSIS_PLUTO_URI ?? '').trim();
  return env !== '' ? env : PLUTO_URI;
}
/** cf-ad9361-dds-core-lpc 16-bit sample scale (12-bit DAC, left room for no clip). */
export const PLUTO_TX_SCALE = 8192;

/** Convert offset-binary u8 IQ to little-endian interleaved int16 (I,Q,...). */
export function u8IqToInt16Le(iqU8: Uint8Array, scale = PLUTO_TX_SCALE): Buffer {
  if (iqU8.length % 2 !== 0) throw new Error('u8IqToInt16Le: odd IQ length');
  const pairs = iqU8.length / 2;
  const buf = Buffer.alloc(pairs * 4); // 2 int16 per pair
  for (let p = 0; p < pairs; p++) {
    const { re, im } = complexFromU8Pair(iqU8[2 * p]!, iqU8[2 * p + 1]!);
    const i = Math.max(-32768, Math.min(32767, Math.round(re * scale)));
    const q = Math.max(-32768, Math.min(32767, Math.round(im * scale)));
    buf.writeInt16LE(i, p * 4);
    buf.writeInt16LE(q, p * 4 + 2);
  }
  return buf;
}

/**
 * Handles the gnosis write Int16 Iq File workflow.
 */
export function writeInt16IqFile(path: string, iqU8: Uint8Array, scale = PLUTO_TX_SCALE): number {
  const buf = u8IqToInt16Le(iqU8, scale);
  writeFileSync(path, buf);
  return buf.length;
}

export interface LibiioGuard {
  readonly ok: boolean;
  readonly buildDir: string;
  readonly framework: string;
  readonly reason?: string;
}

/** Refuse real TX if the ephemeral /tmp libiio build is gone. */
export function checkLibiio(buildDir = LIBIIO_BUILD_DIR): LibiioGuard {
  const framework = `${buildDir}/iio.framework/iio`;
  if (!existsSync(framework)) {
    return {
      ok: false,
      buildDir,
      framework,
      reason: `libiio framework not found at ${framework} (/tmp is ephemeral). Rebuild libiio v0.25 before real TX.`,
    };
  }
  return { ok: true, buildDir, framework };
}

/** iio_attr command argv lists to set TX LO, sample rate, and TX gain. */
export function buildPlutoSetupCmds(opts: {
  uri?: string;
  txLoHz: number;
  sampleRateHz: number;
  hardwaregainDb: number;
  rfBandwidthHz?: number;
}): string[][] {
  const uri = opts.uri ?? resolvePlutoUri();
  const bw = opts.rfBandwidthHz ?? Math.max(200_000, Math.round(opts.sampleRateHz));
  return [
    ['-u', uri, '-c', 'ad9361-phy', 'altvoltage1', 'frequency', String(Math.round(opts.txLoHz))],
    ['-u', uri, '-c', 'ad9361-phy', 'voltage0', 'sampling_frequency', String(Math.round(opts.sampleRateHz))],
    ['-u', uri, '-c', 'ad9361-phy', 'voltage0', 'rf_bandwidth', String(Math.round(bw))],
    ['-u', uri, '-c', 'ad9361-phy', 'voltage0', 'hardwaregain', String(opts.hardwaregainDb)],
  ];
}

/** iio_writedev argv to cyclically stream the int16 IQ file to the TX DAC. */
export function buildIioWritedevArgv(opts: {
  uri?: string;
  bufferSize: number;
  cyclic?: boolean;
}): string[] {
  const uri = opts.uri ?? resolvePlutoUri();
  const argv = ['-u', uri, '-b', String(opts.bufferSize)];
  if (opts.cyclic !== false) argv.push('-c');
  argv.push('cf-ad9361-dds-core-lpc', 'voltage0', 'voltage1');
  return argv;
}
