// @a0n/.ignored_gnosis/rtlsdr-mock-sim/rf-cache-key.ts
import { spawnSync } from 'node:child_process';
import dgram from 'node:dgram';
import { mkdirSync, readFileSync, statSync } from 'node:fs';
import { dirname, resolve } from 'node:path';
import {
  MONSTER_NUMBER_SYMBOL_MAGIC,
  buildProtocol69ProjectionEnvelope,
  type Protocol69ProjectionEnvelope,
} from '../src/protocol69.js';

export const RF_CACHE_KEY_SCHEMA_VERSION = 'gnosis.rf-cache-key.v1';
export const RF_CACHE_KEY_FRAME_MAGIC = 'GNOSIS-RFKEY/1';
export const PROTOCOL69_APP_CACHE_KEY_FRAME_MAGIC = 'GNOSIS-P69-APP/1';
export const PROTOCOL69_FOIL_ADMISSION_CERTIFICATE_VERSION =
  'gnosis.protocol69.foil-admission.v1';
export type RfSymbolMode = 'cache-key-frame' | 'monster-number' | 'polyglot-app-cache-key';
const FLOW_HEADER_SIZE = 10;
const FLOW_FLAGS_FORK_RACE_FOLD_VENT = 0b0000_1111;

export interface RfCacheKeyPacket {
  readonly prefixHash: number;
  readonly prefixLen: number;
  readonly boundaryDimension: number;
  readonly stationCount: number;
}

export interface RfCacheKeyFrame {
  readonly magic: typeof RF_CACHE_KEY_FRAME_MAGIC;
  readonly packet: RfCacheKeyPacket;
  readonly crc16: number;
  readonly ascii: string;
}

export type MonsterNumberProjection = Protocol69ProjectionEnvelope;

export interface RtlSdrDevice {
  readonly index: number;
  readonly vendor: string;
  readonly product: string;
  readonly serial: string;
}

export interface RtlSdrProbe {
  readonly binary: string;
  readonly devices: readonly RtlSdrDevice[];
  readonly healthy: boolean;
  readonly status: number | null;
  readonly stderr: string;
}

export type ExternalTxKind = 'executable-stdin' | 'hackrf-transfer' | 'soapy-sdr-tx';

export interface ExternalTxConfig {
  readonly kind: ExternalTxKind;
  readonly binary: string;
  readonly centerFrequencyHz: number;
  readonly sampleRateHz: number;
  readonly gainDb: number;
  readonly iqFile?: string;
  readonly deviceArgs?: string;
}

export interface ExternalTxAdapter {
  readonly kind: ExternalTxKind;
  readonly binary: string;
  readonly available: boolean;
  readonly argv: readonly string[];
  readonly centerFrequencyHz: number;
  readonly sampleRateHz: number;
  readonly gainDb: number;
  readonly stdinFrame?: string;
  readonly reason?: string;
  readonly discovery?: ExternalTxDiscovery;
}

export interface ExternalTxExecution {
  readonly attempted: boolean;
  readonly status: number | null;
  readonly signal: string | null;
  readonly stdout: string;
  readonly stderr: string;
}

export interface ExternalTxDiscovery {
  readonly attempted: boolean;
  readonly binary: string;
  readonly argv: readonly string[];
  readonly hardwarePresent: boolean;
  readonly status: number | null;
  readonly stdout: string;
  readonly stderr: string;
  readonly error?: string;
}

export interface RtlSdrCaptureConfig {
  readonly binary: string;
  readonly outputPath: string;
  readonly centerFrequencyHz: number;
  readonly sampleRateHz: number;
  readonly samplePairs: number;
  readonly deviceIndex?: string;
}

export interface RtlSdrCaptureResult {
  readonly attempted: boolean;
  readonly binary: string;
  readonly argv: readonly string[];
  readonly outputPath: string;
  readonly centerFrequencyHz: number;
  readonly sampleRateHz: number;
  readonly samplePairs: number;
  readonly ok: boolean;
  readonly status: number | null;
  readonly stdout: string;
  readonly stderr: string;
  readonly byteLength?: number;
  readonly energy?: RtlSdrEnergyStats;
  readonly capturedAtIso: string;
  readonly error?: string;
}

export interface RtlSdrEnergyStats {
  readonly iqPairs: number;
  readonly meanI: number;
  readonly meanQ: number;
  readonly meanMagnitudeSquared: number;
  readonly rmsMagnitude: number;
  readonly peakMagnitude: number;
  readonly dcOffsetMagnitude: number;
  readonly energyDetected: boolean;
}

export interface MonsterRfCacheKeyEnvelope {
  readonly monsterRfCacheKey: {
    readonly schemaVersion: typeof RF_CACHE_KEY_SCHEMA_VERSION;
    readonly transport: 'simulated' | 'hardware';
    readonly broadcaster: {
      readonly kind: 'simulated-baseband' | 'external-tx';
      readonly txAvailable: boolean;
      readonly reason?: string;
      readonly adapter?: ExternalTxAdapter;
    };
    readonly receiver: {
      readonly kind: 'rtl-sdr' | 'simulated-baseband';
      readonly deviceSerials: readonly string[];
      readonly healthy: boolean;
      readonly capture?: RtlSdrCaptureResult;
    };
    readonly frame: RfCacheKeyFrame;
    readonly projection?: MonsterNumberProjection;
    readonly appTeleport?: Protocol69AppCacheKeyTransmission;
    readonly protocol69FoilProbe?: Protocol69FoilProbeReport;
    readonly protocol69FoilAdmission?: Protocol69FoilAdmissionCertificate;
    readonly replayContract: {
      readonly prefixHash: 66;
      readonly prefixLen: 7;
      readonly cachedReplay: true;
      readonly geodesicLength: 0;
    };
  };
}

export interface Protocol69AppCacheKeyReceiptLike {
  readonly schemaVersion: string;
  readonly teleportKind: string;
  readonly receiptReturn: string;
  readonly keyMaterial: {
    readonly leanModule?: string;
    readonly language?: string;
    readonly filePath?: string;
    readonly functionName?: string;
    readonly programChecksum?: string;
  };
  readonly computeCacheKey: string;
}

export interface Protocol69AppCacheKeyFrame {
  readonly magic: typeof PROTOCOL69_APP_CACHE_KEY_FRAME_MAGIC;
  readonly receiptSchemaVersion: string;
  readonly computeCacheKey: string;
  readonly programChecksum?: string;
  readonly functionName?: string;
  readonly crc16: number;
  readonly ascii: string;
}

export interface Protocol69AppCacheKeyTransmission {
  readonly receipt: Protocol69AppCacheKeyReceiptLike;
  readonly frame: Protocol69AppCacheKeyFrame;
}

export type Protocol69FoilProbeTransport = 'http' | 'flow-udp' | 'both';

export interface Protocol69FoilProbeConfig {
  readonly transport: Protocol69FoilProbeTransport;
  readonly url: string;
  readonly flowHost: string;
  readonly flowPort: number;
  readonly flowPath: string;
  readonly timeoutMs: number;
}

export interface Protocol69FoilProbeLeg {
  readonly attempted: boolean;
  readonly ok: boolean;
  readonly status: number | null;
  readonly endpoint: string;
  readonly response?: unknown;
  readonly error?: string;
}

export interface Protocol69FoilProbeReport {
  readonly protocol: 'protocol69';
  readonly ok: boolean;
  readonly transport: Protocol69FoilProbeTransport;
  readonly http?: Protocol69FoilProbeLeg;
  readonly flowUdp?: Protocol69FoilProbeLeg;
}

export interface Protocol69FoilAdmissionCertificate {
  readonly certificateVersion: typeof PROTOCOL69_FOIL_ADMISSION_CERTIFICATE_VERSION;
  readonly protocol: 'protocol69';
  readonly wordName: 'sixty9';
  readonly source: 'monster-rf-cache-key';
  readonly projectionAccepted: boolean;
  readonly requiredTransport: Protocol69FoilProbeTransport;
  readonly httpAccepted?: boolean;
  readonly flowUdpAccepted?: boolean;
  readonly broadcastSymbol: 66;
  readonly localFoilOperand: 7;
  readonly expectedProjection: 69;
  readonly theoremRefs: readonly string[];
}

export const canonicalRfCacheKeyPacket: RfCacheKeyPacket = {
  prefixHash: 66,
  prefixLen: 7,
  boundaryDimension: 12,
  stationCount: 2,
};

/**
 * Handles the gnosis encode Monster Number Projection workflow.
 */
export function encodeMonsterNumberProjection(): MonsterNumberProjection {
  return buildProtocol69ProjectionEnvelope();
}

/**
 * Builds the Protocol69 Foil Admission Certificate.
 */
export function buildProtocol69FoilAdmissionCertificate(
  report?: Protocol69FoilProbeReport,
): Protocol69FoilAdmissionCertificate | undefined {
  if (!report) {
    return undefined;
  }

  return {
    certificateVersion: PROTOCOL69_FOIL_ADMISSION_CERTIFICATE_VERSION,
    protocol: 'protocol69',
    wordName: 'sixty9',
    source: 'monster-rf-cache-key',
    projectionAccepted: report.ok,
    requiredTransport: report.transport,
    httpAccepted: report.http?.ok,
    flowUdpAccepted: report.flowUdp?.ok,
    broadcastSymbol: 66,
    localFoilOperand: 7,
    expectedProjection: 69,
    theoremRefs: [
      'Gnosis.RemoteCacheKeyTeleportation.protocol69_monster_number_projection',
      'Gnosis.RemoteCacheKeyTeleportation.protocol69_braids_integer_projection_with_fano_parity',
    ],
  };
}

function validateProtocol69ProjectionResponse(body: unknown): string | null {
  if (typeof body !== 'object' || body === null || !('projection' in body)) {
    return 'response missing projection';
  }
  const projection = (body as { readonly projection?: unknown }).projection;
  if (typeof projection !== 'object' || projection === null) {
    return 'response projection is not an object';
  }
  const record = projection as Partial<MonsterNumberProjection>;
  if (record.protocol !== 'protocol69') return 'protocol mismatch';
  if (record.wordName !== 'sixty9') return 'wordName mismatch';
  if (record.broadcastSymbol !== 66) return 'broadcastSymbol mismatch';
  if (record.localFoilOperand !== 7) return 'localFoilOperand mismatch';
  if (record.expectedProjection !== 69) return 'expectedProjection mismatch';
  return null;
}

function encodeFlowFrame(payload: string, sequence = 0): Buffer {
  const payloadBytes = Buffer.from(payload);
  if (payloadBytes.length > 0x00ff_ffff) {
    throw new Error('payload exceeds 24-bit Aeon Flow frame length');
  }
  const frame = Buffer.alloc(FLOW_HEADER_SIZE + payloadBytes.length);
  frame.writeUInt16BE(1, 0);
  frame.writeUInt32BE(sequence, 2);
  frame[6] = FLOW_FLAGS_FORK_RACE_FOLD_VENT;
  frame[7] = (payloadBytes.length >> 16) & 0xff;
  frame[8] = (payloadBytes.length >> 8) & 0xff;
  frame[9] = payloadBytes.length & 0xff;
  payloadBytes.copy(frame, FLOW_HEADER_SIZE);
  return frame;
}

function decodeFlowFramePayload(bytes: Buffer): string {
  if (bytes.length < FLOW_HEADER_SIZE) {
    throw new Error('buffer smaller than Aeon Flow header');
  }
  const payloadLength = (bytes[7]! << 16) | (bytes[8]! << 8) | bytes[9]!;
  if (bytes.length < FLOW_HEADER_SIZE + payloadLength) {
    throw new Error('incomplete Aeon Flow payload');
  }
  return bytes.subarray(FLOW_HEADER_SIZE, FLOW_HEADER_SIZE + payloadLength).toString('utf8');
}

function parseJsonMaybe(text: string): unknown {
  if (text.trim() === '') return '';
  try {
    return JSON.parse(text) as unknown;
  } catch {
    return text;
  }
}

/**
 * Handles the gnosis probe Protocol69 Foil Http workflow.
 */
export async function probeProtocol69FoilHttp(
  url: string,
  timeoutMs: number,
): Promise<Protocol69FoilProbeLeg> {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), timeoutMs);
  try {
    const response = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(encodeMonsterNumberProjection()),
      signal: controller.signal,
    });
    const body = parseJsonMaybe(await response.text());
    const validationError = response.ok ? validateProtocol69ProjectionResponse(body) : null;
    return {
      attempted: true,
      ok: response.ok && validationError === null,
      status: response.status,
      endpoint: url,
      response: body,
      error: validationError ?? undefined,
    };
  } catch (error) {
    return {
      attempted: true,
      ok: false,
      status: null,
      endpoint: url,
      error: error instanceof Error ? error.message : String(error),
    };
  } finally {
    clearTimeout(timeout);
  }
}

/**
 * Handles the gnosis probe Protocol69 Foil Flow Udp workflow.
 */
export async function probeProtocol69FoilFlowUdp(opts: {
  readonly host: string;
  readonly port: number;
  readonly path: string;
  readonly timeoutMs: number;
}): Promise<Protocol69FoilProbeLeg> {
  const endpoint = `aeon://${opts.host}:${opts.port}${opts.path}`;
  const socket = dgram.createSocket('udp4');
  try {
    const frame = encodeFlowFrame(opts.path);
    const response = await new Promise<Buffer>((resolve, reject) => {
      const timeout = setTimeout(() => {
        reject(new Error(`Flow/UDP timeout after ${opts.timeoutMs}ms`));
      }, opts.timeoutMs);
      socket.once('message', (message: Buffer) => {
        clearTimeout(timeout);
        resolve(message);
      });
      socket.once('error', (error: Error) => {
        clearTimeout(timeout);
        reject(error);
      });
      socket.send(frame, opts.port, opts.host);
    });
    const body = parseJsonMaybe(decodeFlowFramePayload(response));
    const validationError = validateProtocol69ProjectionResponse(body);
    return {
      attempted: true,
      ok: validationError === null,
      status: validationError === null ? 200 : 502,
      endpoint,
      response: body,
      error: validationError ?? undefined,
    };
  } catch (error) {
    return {
      attempted: true,
      ok: false,
      status: null,
      endpoint,
      error: error instanceof Error ? error.message : String(error),
    };
  } finally {
    socket.close();
  }
}

/**
 * Handles the gnosis probe Protocol69 Foil workflow.
 */
export async function probeProtocol69Foil(
  config: Protocol69FoilProbeConfig,
): Promise<Protocol69FoilProbeReport> {
  const http =
    config.transport === 'http' || config.transport === 'both'
      ? await probeProtocol69FoilHttp(config.url, config.timeoutMs)
      : undefined;
  const flowUdp =
    config.transport === 'flow-udp' || config.transport === 'both'
      ? await probeProtocol69FoilFlowUdp({
          host: config.flowHost,
          port: config.flowPort,
          path: config.flowPath,
          timeoutMs: config.timeoutMs,
        })
      : undefined;
  return {
    protocol: 'protocol69',
    transport: config.transport,
    ok: (http?.ok ?? true) && (flowUdp?.ok ?? true),
    http,
    flowUdp,
  };
}

export const defaultRfTxCenterFrequencyHz = 433_920_000;
export const defaultRfTxSampleRateHz = 2_000_000;
export const defaultRfTxGainDb = 0;
export const defaultRfKeyRxSamplePairs = 409_600;
export const defaultRfKeyRxCapturePath = '.qa-artifacts/rtlsdr/rf-key-window.u8';

/**
 * Handles the gnosis crc16 Ccitt Ascii workflow.
 */
export function crc16CcittAscii(input: string): number {
  let crc = 0xffff;
  for (let i = 0; i < input.length; i++) {
    crc ^= input.charCodeAt(i) << 8;
    for (let bit = 0; bit < 8; bit++) {
      crc = (crc & 0x8000) !== 0 ? ((crc << 1) ^ 0x1021) : (crc << 1);
      crc &= 0xffff;
    }
  }
  return crc;
}

/**
 * Handles the gnosis encode Rf Cache Key Frame workflow.
 */
export function encodeRfCacheKeyFrame(packet: RfCacheKeyPacket): RfCacheKeyFrame {
  const body = [
    RF_CACHE_KEY_FRAME_MAGIC,
    `prefixHash=${packet.prefixHash}`,
    `prefixLen=${packet.prefixLen}`,
    `boundaryDimension=${packet.boundaryDimension}`,
    `stationCount=${packet.stationCount}`,
  ].join(' ');
  const crc16 = crc16CcittAscii(body);
  return {
    magic: RF_CACHE_KEY_FRAME_MAGIC,
    packet,
    crc16,
    ascii: `${body} crc16=${crc16.toString(16).padStart(4, '0')}`,
  };
}

/**
 * Handles the gnosis encode Protocol69 App Cache Key Frame workflow.
 */
export function encodeProtocol69AppCacheKeyFrame(
  receipt: Protocol69AppCacheKeyReceiptLike,
): Protocol69AppCacheKeyFrame {
  const parts = [
    PROTOCOL69_APP_CACHE_KEY_FRAME_MAGIC,
    `schema=${receipt.schemaVersion}`,
    `cacheKey=${receipt.computeCacheKey}`,
  ];
  if (receipt.keyMaterial.functionName) {
    parts.push(`function=${receipt.keyMaterial.functionName}`);
  }
  if (receipt.keyMaterial.programChecksum) {
    parts.push(`programChecksum=${receipt.keyMaterial.programChecksum}`);
  }
  const body = parts.join(' ');
  const crc16 = crc16CcittAscii(body);
  return {
    magic: PROTOCOL69_APP_CACHE_KEY_FRAME_MAGIC,
    receiptSchemaVersion: receipt.schemaVersion,
    computeCacheKey: receipt.computeCacheKey,
    programChecksum: receipt.keyMaterial.programChecksum,
    functionName: receipt.keyMaterial.functionName,
    crc16,
    ascii: `${body} crc16=${crc16.toString(16).padStart(4, '0')}`,
  };
}

/**
 * Builds the Protocol69 App Cache Key Transmission.
 */
export function buildProtocol69AppCacheKeyTransmission(
  receipt: Protocol69AppCacheKeyReceiptLike,
): Protocol69AppCacheKeyTransmission {
  if (receipt.receiptReturn !== 'compute-cache-key') {
    throw new Error('protocol69 app receipt must return compute-cache-key');
  }
  if (!receipt.computeCacheKey.startsWith('gnosis:protocol69:polyglot-app-cache:v1:')) {
    throw new Error('protocol69 app receipt has an unsupported compute cache key prefix');
  }
  return {
    receipt,
    frame: encodeProtocol69AppCacheKeyFrame(receipt),
  };
}

/**
 * Parses the Rtl Test Devices.
 */
export function parseRtlTestDevices(stderr: string): RtlSdrDevice[] {
  const devices: RtlSdrDevice[] = [];
  const re = /^\s*(\d+):\s*([^,]+),\s*([^,]+),\s*SN:\s*(\S+)\s*$/gm;
  let match: RegExpExecArray | null;
  while ((match = re.exec(stderr)) !== null) {
    devices.push({
      index: Number.parseInt(match[1]!, 10),
      vendor: match[2]!.trim(),
      product: match[3]!.trim(),
      serial: match[4]!.trim(),
    });
  }
  return devices;
}

/**
 * Handles the gnosis probe Rtl Sdr workflow.
 */
export function probeRtlSdr(binary = 'rtl_test'): RtlSdrProbe {
  const result = spawnSync(binary, ['-t'], {
    encoding: 'utf8',
    timeout: 10_000,
  });
  const stderr = `${result.stderr ?? ''}${result.stdout ?? ''}`;
  const devices = parseRtlTestDevices(stderr);
  return {
    binary,
    devices,
    healthy: (result.status ?? 1) === 0 && devices.length > 0,
    status: result.status,
    stderr,
  };
}

/**
 * Handles the gnosis capture Rtl Sdr Window workflow.
 */
export function captureRtlSdrWindow(
  config: RtlSdrCaptureConfig,
): RtlSdrCaptureResult {
  const outputPath = resolve(config.outputPath);
  mkdirSync(dirname(outputPath), { recursive: true });
  const argv = [
    '-f',
    String(config.centerFrequencyHz),
    '-s',
    String(config.sampleRateHz),
    '-n',
    String(config.samplePairs),
  ];
  if (config.deviceIndex && config.deviceIndex.trim() !== '') {
    argv.push('-d', config.deviceIndex);
  }
  argv.push(outputPath);
  const result = spawnSync(config.binary, argv, {
    encoding: 'utf8',
    timeout: 30_000,
  });
  const statusOk = (result.status ?? 1) === 0;
  let byteLength: number | undefined;
  let energy: RtlSdrEnergyStats | undefined;
  if (statusOk) {
    try {
      byteLength = statSync(outputPath).size;
      energy = analyzeRtlU8Energy(readFileSync(outputPath));
    } catch {
      byteLength = undefined;
      energy = undefined;
    }
  }
  const expectedBytes = config.samplePairs * 2;
  const ok = statusOk && byteLength === expectedBytes;
  return {
    attempted: true,
    binary: config.binary,
    argv,
    outputPath,
    centerFrequencyHz: config.centerFrequencyHz,
    sampleRateHz: config.sampleRateHz,
    samplePairs: config.samplePairs,
    ok,
    status: result.status,
    stdout: result.stdout ?? '',
    stderr: result.stderr ?? '',
    byteLength,
    energy,
    capturedAtIso: new Date().toISOString(),
    error: result.error?.message,
  };
}

/**
 * Handles the gnosis analyze Rtl U8 Energy workflow.
 */
export function analyzeRtlU8Energy(iqU8: Uint8Array): RtlSdrEnergyStats {
  if (iqU8.length < 2 || iqU8.length % 2 !== 0) {
    return {
      iqPairs: 0,
      meanI: 0,
      meanQ: 0,
      meanMagnitudeSquared: 0,
      rmsMagnitude: 0,
      peakMagnitude: 0,
      dcOffsetMagnitude: 0,
      energyDetected: false,
    };
  }
  const iqPairs = iqU8.length / 2;
  let sumI = 0;
  let sumQ = 0;
  let sumMag2 = 0;
  let peakMag2 = 0;
  for (let p = 0; p + 1 < iqU8.length; p += 2) {
    const i = (iqU8[p]! - 128) / 128;
    const q = (iqU8[p + 1]! - 128) / 128;
    const mag2 = i * i + q * q;
    sumI += i;
    sumQ += q;
    sumMag2 += mag2;
    if (mag2 > peakMag2) peakMag2 = mag2;
  }
  const meanI = sumI / iqPairs;
  const meanQ = sumQ / iqPairs;
  const meanMagnitudeSquared = sumMag2 / iqPairs;
  const rmsMagnitude = Math.sqrt(meanMagnitudeSquared);
  const peakMagnitude = Math.sqrt(peakMag2);
  const dcOffsetMagnitude = Math.sqrt(meanI * meanI + meanQ * meanQ);
  return {
    iqPairs,
    meanI,
    meanQ,
    meanMagnitudeSquared,
    rmsMagnitude,
    peakMagnitude,
    dcOffsetMagnitude,
    energyDetected: rmsMagnitude > 0.02 || peakMagnitude > 0.08,
  };
}

/**
 * Handles the gnosis infer External Tx Kind workflow.
 */
export function inferExternalTxKind(binary: string): ExternalTxKind {
  const lower = binary.toLowerCase();
  if (lower.includes('hackrf_transfer')) return 'hackrf-transfer';
  if (lower.includes('soapysdr')) return 'soapy-sdr-tx';
  return 'executable-stdin';
}

/**
 * Handles the gnosis probe Soapy Sdr Device workflow.
 */
export function probeSoapySdrDevice(
  binary: string,
  deviceArgs: string,
): ExternalTxDiscovery {
  const argv = deviceArgs.trim() === '' ? ['--find'] : [`--find=${deviceArgs}`];
  const result = spawnSync(binary, argv, {
    encoding: 'utf8',
    timeout: 10_000,
  });
  const stdout = result.stdout ?? '';
  const stderr = result.stderr ?? '';
  const combined = `${stdout}\n${stderr}`.toLowerCase();
  const hardwarePresent =
    (result.status ?? 1) === 0 &&
    combined.trim() !== '' &&
    !combined.includes('no devices found') &&
    !combined.includes('found device(s): 0');
  return {
    attempted: true,
    binary,
    argv,
    hardwarePresent,
    status: result.status,
    stdout,
    stderr,
    error: result.error?.message,
  };
}

/**
 * Builds the External Tx Adapter.
 */
export function buildExternalTxAdapter(
    config: ExternalTxConfig,
    stdinFrame: string,
    discovery?: ExternalTxDiscovery,
): ExternalTxAdapter {
  if (config.binary.trim() === '') {
    return {
      kind: config.kind,
      binary: config.binary,
      available: false,
      argv: [],
      centerFrequencyHz: config.centerFrequencyHz,
      sampleRateHz: config.sampleRateHz,
      gainDb: config.gainDb,
      reason: 'no transmitter binary configured',
      discovery,
    };
  }

  if (config.kind === 'hackrf-transfer') {
    if (!config.iqFile || config.iqFile.trim() === '') {
      return {
        kind: config.kind,
        binary: config.binary,
        available: false,
        argv: [],
        centerFrequencyHz: config.centerFrequencyHz,
        sampleRateHz: config.sampleRateHz,
        gainDb: config.gainDb,
        reason: 'hackrf_transfer requires a premodulated IQ file via --tx-iq-file or GNOSIS_RF_TX_IQ_FILE',
        discovery,
      };
    }
    return {
      kind: config.kind,
      binary: config.binary,
      available: true,
      argv: [
        '-t',
        config.iqFile,
        '-f',
        String(config.centerFrequencyHz),
        '-s',
        String(config.sampleRateHz),
        '-x',
        String(config.gainDb),
      ],
      centerFrequencyHz: config.centerFrequencyHz,
      sampleRateHz: config.sampleRateHz,
      gainDb: config.gainDb,
      discovery,
    };
  }

  if (config.kind === 'soapy-sdr-tx') {
    if (!config.deviceArgs || config.deviceArgs.trim() === '') {
      return {
        kind: config.kind,
        binary: config.binary,
        available: false,
        argv: [],
        centerFrequencyHz: config.centerFrequencyHz,
        sampleRateHz: config.sampleRateHz,
        gainDb: config.gainDb,
        reason: 'SoapySDR TX requires device arguments via --tx-device-args or GNOSIS_RF_TX_DEVICE_ARGS',
        discovery,
      };
    }
    if (discovery?.attempted === true && !discovery.hardwarePresent) {
      return {
        kind: config.kind,
        binary: config.binary,
        available: false,
        argv: [],
        centerFrequencyHz: config.centerFrequencyHz,
        sampleRateHz: config.sampleRateHz,
        gainDb: config.gainDb,
        reason: discovery.error ?? 'SoapySDR discovery did not find the requested transmitter',
        discovery,
      };
    }
    return {
      kind: config.kind,
      binary: config.binary,
      available: true,
      argv: [
        '--device',
        config.deviceArgs,
        '--rate',
        String(config.sampleRateHz),
        '--freq',
        String(config.centerFrequencyHz),
        '--gain',
        String(config.gainDb),
        '--stdin-ascii-frame',
      ],
      centerFrequencyHz: config.centerFrequencyHz,
      sampleRateHz: config.sampleRateHz,
      gainDb: config.gainDb,
      stdinFrame,
      discovery,
    };
  }

  return {
    kind: config.kind,
    binary: config.binary,
    available: true,
    argv: [
      '--center-frequency-hz',
      String(config.centerFrequencyHz),
      '--sample-rate-hz',
      String(config.sampleRateHz),
      '--gain-db',
      String(config.gainDb),
    ],
    centerFrequencyHz: config.centerFrequencyHz,
    sampleRateHz: config.sampleRateHz,
    gainDb: config.gainDb,
    stdinFrame,
    discovery,
  };
}

/**
 * Handles the gnosis execute External Tx Adapter workflow.
 */
export function executeExternalTxAdapter(
  adapter: ExternalTxAdapter,
): ExternalTxExecution {
  if (!adapter.available) {
    return {
      attempted: false,
      status: null,
      signal: null,
      stdout: '',
      stderr: adapter.reason ?? 'transmitter adapter is not available',
    };
  }
  const result = spawnSync(adapter.binary, [...adapter.argv], {
    encoding: 'utf8',
    input: adapter.stdinFrame,
    timeout: 30_000,
  });
  return {
    attempted: true,
    status: result.status,
    signal: result.signal,
    stdout: result.stdout ?? '',
    stderr: result.stderr ?? '',
  };
}

/**
 * Builds the Monster Rf Cache Key Envelope.
 */
export function buildMonsterRfCacheKeyEnvelope(opts: {
  readonly transport: 'simulated' | 'hardware';
  readonly rtlProbe?: RtlSdrProbe;
  readonly txBin?: string | null;
  readonly txKind?: ExternalTxKind | null;
  readonly txCenterFrequencyHz?: number;
  readonly txSampleRateHz?: number;
  readonly txGainDb?: number;
  readonly txIqFile?: string | null;
  readonly txDeviceArgs?: string | null;
  readonly txDiscovery?: ExternalTxDiscovery;
  readonly rxCapture?: RtlSdrCaptureResult;
  readonly symbolMode?: RfSymbolMode;
  readonly appTeleportReceipt?: Protocol69AppCacheKeyReceiptLike;
  readonly protocol69FoilProbe?: Protocol69FoilProbeReport;
}): MonsterRfCacheKeyEnvelope {
  const frame = encodeRfCacheKeyFrame(canonicalRfCacheKeyPacket);
  const projection =
    opts.symbolMode === 'monster-number' ? encodeMonsterNumberProjection() : undefined;
  const appTeleport =
    opts.symbolMode === 'polyglot-app-cache-key' && opts.appTeleportReceipt
      ? buildProtocol69AppCacheKeyTransmission(opts.appTeleportReceipt)
      : undefined;
  const stdinFrame = appTeleport?.frame.ascii ?? projection?.ascii ?? frame.ascii;
  const deviceSerials = opts.rtlProbe?.devices.map((d) => d.serial) ?? [];
  const txBin = opts.txBin?.trim() ?? '';
  const adapter =
    opts.transport === 'hardware' && txBin !== ''
      ? buildExternalTxAdapter(
          {
            kind: opts.txKind ?? inferExternalTxKind(txBin),
            binary: txBin,
            centerFrequencyHz:
              opts.txCenterFrequencyHz ?? defaultRfTxCenterFrequencyHz,
            sampleRateHz: opts.txSampleRateHz ?? defaultRfTxSampleRateHz,
            gainDb: opts.txGainDb ?? defaultRfTxGainDb,
            iqFile: opts.txIqFile ?? undefined,
            deviceArgs: opts.txDeviceArgs ?? undefined,
          },
          stdinFrame,
          opts.txDiscovery,
        )
      : undefined;
  const txAvailable = opts.transport === 'simulated' || (adapter?.available ?? false);
  const protocol69FoilAdmission = buildProtocol69FoilAdmissionCertificate(
    opts.protocol69FoilProbe,
  );
  return {
    monsterRfCacheKey: {
      schemaVersion: RF_CACHE_KEY_SCHEMA_VERSION,
      transport: opts.transport,
      broadcaster:
        opts.transport === 'simulated'
          ? { kind: 'simulated-baseband', txAvailable: true }
          : {
              kind: 'external-tx',
              txAvailable,
              reason: txAvailable
                ? undefined
                : adapter?.reason ??
                  'RTL-SDR devices are receive-only; set GNOSIS_RF_TX_BIN or --tx-bin for live RF transmit.',
              adapter,
            },
      receiver:
        opts.transport === 'simulated'
          ? { kind: 'simulated-baseband', deviceSerials: [], healthy: true }
          : {
              kind: 'rtl-sdr',
              deviceSerials,
              healthy: opts.rxCapture?.ok ?? opts.rtlProbe?.healthy ?? false,
              capture: opts.rxCapture,
            },
      frame,
      projection,
      appTeleport,
      protocol69FoilProbe: opts.protocol69FoilProbe,
      protocol69FoilAdmission,
      replayContract: {
        prefixHash: 66,
        prefixLen: 7,
        cachedReplay: true,
        geodesicLength: 0,
      },
    },
  };
}
