// @a0n/.ignored_gnosis/rtlsdr-mock-sim/rtl433.ts
import { readFileSync } from 'node:fs';

export interface Rtl433Event {
  readonly whenIso: string | null;
  readonly model: string;
  readonly frequencyHz: number | null;
  readonly id: string | null;
  readonly channel: string | null;
  readonly raw: Record<string, unknown>;
}

function asRecord(v: unknown): Record<string, unknown> | null {
  return typeof v === 'object' && v !== null && !Array.isArray(v)
    ? (v as Record<string, unknown>)
    : null;
}

function readString(
  obj: Record<string, unknown>,
  keys: readonly string[],
): string | null {
  for (const k of keys) {
    const v = obj[k];
    if (typeof v === 'string' && v.trim() !== '') return v.trim();
    if (typeof v === 'number' && Number.isFinite(v)) return String(v);
  }
  return null;
}

function readNumber(
  obj: Record<string, unknown>,
  keys: readonly string[],
): number | null {
  for (const k of keys) {
    const v = obj[k];
    if (typeof v === 'number' && Number.isFinite(v)) return v;
    if (typeof v === 'string' && v.trim() !== '') {
      const n = Number.parseFloat(v);
      if (Number.isFinite(n)) return n;
    }
  }
  return null;
}

/**
 * Parses the Rtl433 Json Line.
 */
export function parseRtl433JsonLine(line: string): Rtl433Event | null {
  const trimmed = line.trim();
  if (trimmed === '') return null;
  let parsed: unknown;
  try {
    parsed = JSON.parse(trimmed);
  } catch {
    return null;
  }
  const obj = asRecord(parsed);
  if (obj === null) return null;
  const model = readString(obj, ['model', 'type', 'protocol']) ?? 'unknown';
  const whenIso = readString(obj, ['time', 'datetime', 'timestamp']);
  const frequencyHz = readNumber(obj, ['freq', 'frequency', 'frequency_hz']);
  const id = readString(obj, ['id', 'device_id', 'sensor_id']);
  const channel = readString(obj, ['channel', 'ch']);
  return {
    whenIso,
    model,
    frequencyHz,
    id,
    channel,
    raw: obj,
  };
}

/**
 * Loads the Rtl433 Jsonl.
 */
export function loadRtl433Jsonl(path: string): Rtl433Event[] {
  const text = readFileSync(path, 'utf8');
  const out: Rtl433Event[] = [];
  for (const line of text.split(/\r?\n/)) {
    const ev = parseRtl433JsonLine(line);
    if (ev !== null) out.push(ev);
  }
  return out;
}

/**
 * Handles the gnosis summarize Rtl433 Events workflow.
 */
export function summarizeRtl433Events(events: readonly Rtl433Event[]): {
  readonly count: number;
  readonly byModel: Readonly<Record<string, number>>;
  readonly byRoundedFreqMHz: Readonly<Record<string, number>>;
} {
  const byModel: Record<string, number> = {};
  const byRoundedFreqMHz: Record<string, number> = {};
  for (const ev of events) {
    byModel[ev.model] = (byModel[ev.model] ?? 0) + 1;
    if (ev.frequencyHz !== null) {
      const mhz = (ev.frequencyHz / 1e6).toFixed(3);
      byRoundedFreqMHz[mhz] = (byRoundedFreqMHz[mhz] ?? 0) + 1;
    }
  }
  return {
    count: events.length,
    byModel,
    byRoundedFreqMHz,
  };
}
