// @a0n/.ignored_gnosis/rtlsdr-mock-sim/skymesh-bridge.mts
#!/usr/bin/env node
/**
 * Skymesh hosted-edge bridge client.
 *
 * The inverse of the MonitorRelayDO: a small Node 22+ program that connects the
 * LOCAL aeon-monitor (plaintext HTTP on 127.0.0.1:8722) UP to the hosted relay
 * (WSS at wss://skymesh.forkjoin.ai/monitor/bridge). It forwards frames/status/
 * scenes UP and applies control DOWN, so the deployed Skymesh site can show real
 * live monitor data and drive the local radio with no CORS and no mixed content.
 *
 * The local monitor is never reachable from the public edge (127.0.0.1, plaintext);
 * this bridge dials OUT (outbound WSS, no inbound firewall hole) and the relay
 * caches + fans out the frames it pushes.
 *
 * Uplink protocol (FIXED -- byte-compatible with MonitorRelayDO.webSocketMessage,
 * see distributed-inference/docs/MONITOR_BRIDGE_DESIGN.md section 2.3):
 *
 *   UP   { type: "scenes",  scenes }              (on open + reconnect)
 *   UP   { type: "status",  status }              (on open + on change, ~1s poll)
 *   UP   { type: "frame",   envelope }            (per local SSE envelope)
 *   UP   { type: "control-result", id, ok, status, error }   (reply to a control)
 *   UP   { type: "query-result", id, reply, ... }  (reply to a mesh query)
 *   DOWN { type: "control", id, op, args }        (the relay forwards an app op)
 *   DOWN { type: "query", id, prompt?, tokens?, maxTokens }  (a mesh query)
 *
 * Zero heavy deps: the built-in WebSocket global (Node 22+) and global fetch /
 * ReadableStream for the local read surface. A tiny SSE `data:` parser consumes
 * the local stream server-side (NOT the browser EventSource).
 *
 * Usage: bin/skymesh-bridge [--relay wss://skymesh.forkjoin.ai/monitor/bridge]
 *                           [--local http://127.0.0.1:8722]
 *                           [--token $SKYMESH_BRIDGE_TOKEN]
 *                           [--scenes skymesh,spectrum-waterfall,...]
 *                           [--mesh-id skymesh-global] [--node-id $HOSTNAME]
 *                           [--dashrelay wss://dashrelay.com] [--no-relay]
 *
 * GLOBAL PARTYLINE (mesh id + node id): the relay routes per mesh id. Every
 * bridge publishing the default `skymesh-global` mesh joins one global room, so
 * the deployed app sees every node's frames; `--mesh-id <id>` carves a private
 * mesh. The first frame of identity is a `{type:"hello", nodeId}` on the relay
 * WS (and `?node=` on the upgrade); frames the relay returns carry
 * `metadata.nodeId` so the app can attribute each frame to a node.
 *
 * OPTIONAL DashRelay uplink (`--dashrelay <wssBase>`): an ADDITIVE second uplink
 * to a DashRelay room over the aeon-flow endpoint
 * (`<wssBase>/.aeon/flow?room=<meshId>`, opaque-relay mode). The join handshake is
 * the SAME JSON `{type:"join",room,mode:"opaque-relay",clientId}` text frame the
 * `/relay/sync` path uses (RelayRoomDO.handleAeonFlowUpgrade, src/durables/
 * RelayRoomDO.ts:1000 — first message must be a JSON JOIN, parseJoinMessage:289),
 * and the DO replies `{type:"joined"}` (joinPeer:1424) then `{type:"flow_ready"}`
 * (1107). After join, each local monitor envelope is carried as a single aeon
 * Flow frame on the MonitorFlow stream: `FlowCodec.encode(encodeMonitorFlowFrame
 * (envelope))` (binary), broadcast verbatim to the room's other peers
 * (opaque-relay relays binary frames as-is, processPeerFrame:1517 ->
 * broadcastToMode:1697). The bridge can publish to BOTH the hosted relay and
 * DashRelay, or DashRelay-only with `--no-relay`.
 *
 * DashRelay IS deployed at dashrelay.com (NOT the relay.* subdomain). The working
 * WS endpoint is the aeon-flow path `wss://dashrelay.com/.aeon/flow` (handshake
 * 101); the old `/relay/sync` opaque-relay join did NOT complete in probing
 * (HTTP 500 / WS 1006), so this transport now rides the aeon-flow endpoint and the
 * MonitorFlow Flow wire. Unit-tested against a mock aeon-flow relay.
 */

import os from 'node:os';
import path from 'node:path';
import { execFile } from 'node:child_process';
import {
  FlowCodec,
  encodeMonitorFlowFrame,
  type MonitorFlowFrame,
} from '@a0n/aeon/flow';
import { cacheFingerprint48HexSync } from '@a0n/bitwise/cache-fp48';

// --------------------------------------------------------------------------
// CLI
// --------------------------------------------------------------------------

/** Read a `--name value` flag from argv, or return the default. */
export function argValue(argv: readonly string[], name: string, def: string): string {
  const i = argv.indexOf(name);
  if (i === -1 || i + 1 >= argv.length) return def;
  return argv[i + 1]!;
}

/** Read a boolean `--name` flag (presence => true) from argv. */
export function argFlag(argv: readonly string[], name: string): boolean {
  return argv.includes(name);
}

/** The FIXED uplink message types (kept here so tests assert against the contract). */
export const UPLINK_MESSAGE_TYPES = ['scenes', 'status', 'frame', 'control-result', 'preflight-result', 'query-result'] as const;
export const DOWNLINK_MESSAGE_TYPES = ['control', 'preflight', 'query'] as const;

/** The default GLOBAL PARTYLINE mesh id (shared contract with the relay + app). */
export const SKYMESH_GLOBAL_MESH_ID = 'skymesh-global';

// --------------------------------------------------------------------------
// PREFLIGHT PARIS GATE (FIXED contract shared by bridge + relay + app).
//
// A node proves itself with a REAL inference test before joining the global
// mesh: the relay forwards a {type:"preflight",id} DOWN, the bridge spawns the
// VERIFIED paris-probe-rknot prover against its local model, parses the verdict
// from stdout, and replies {type:"preflight-result",...} UP. A PASS sets the
// bridge's internal `admitted` flag (carried in hello/status). The probe is REAL
// inference -- never a mocked verdict in the shipped path.
// --------------------------------------------------------------------------

/** The fixed preflight test id (the Paris quality gate, v1). */
export const PREFLIGHT_TEST_ID = 'paris/v1';
/** Prompt text the preflight asserts. */
export const PREFLIGHT_PROMPT = 'The capital of France is';
/** Tokenized prompt = "The capital of France is" (the prover's --tokens). */
export const PREFLIGHT_TOKENS = [785, 6722, 315, 9625, 374] as const;
/** Expected next token id = " Paris" (the prover's --expected). */
export const PREFLIGHT_EXPECTED_TOKEN = 12095;
/** Expected next token text. */
export const PREFLIGHT_EXPECTED_TEXT = ' Paris';

/** Default model the prover runs (admitted; ~3s real forward pass). */
const DEFAULT_PREFLIGHT_MODEL = '~/.edgework/models/qwen2.5-0.5b-instruct.knot';
/** Default prover binary, relative to the gnosis repo root. */
const DEFAULT_PROBE_REL =
  'open-source/gnosis/distributed-inference/target/release/paris-probe-rknot';

/** The preflight-result the bridge sends UP (FIXED contract). */
export interface PreflightResult {
  type: 'preflight-result';
  id: string;
  pass: boolean;
  token: number;
  text: string;
  logit: number;
  elapsedMs: number;
  model: string;
  detail?: string;
}

// --------------------------------------------------------------------------
// MESH QUERY (FIXED contract shared by bridge + relay + app + MCP).
//
// The relay forwards a {type:"query",id,prompt?,tokens?,maxTokens} DOWN; the
// bridge runs the REAL per-token autoregressive decode against its local
// fat-station (the SAME reliable /decode-next path bin/skymesh-query proves --
// NOT /generate, whose batched decode repeats the last token) and replies
// {type:"query-result",...} UP. The inference is REAL; the reply is never mocked
// in the shipped path. Only the fetch is injectable (the test seam).
// --------------------------------------------------------------------------

/** qwen end-of-text / im_end tokens -- stop decoding on these. */
export const QUERY_STOP_TOKENS = [151643, 151645, 151644] as const;
/** Default number of new tokens to decode when the request omits maxTokens. */
export const QUERY_DEFAULT_MAX_TOKENS = 24;

/** The query-result the bridge sends UP (FIXED contract). */
export interface QueryResult {
  type: 'query-result';
  id: string;
  /** The detokenized reply text (best-effort; '' if detokenize is unavailable). */
  reply: string;
  /** The prompt token ids (echoed; from --tokens or the node's /tokenize). */
  promptTokens: number[];
  /** The generated token ids (argmax per position, before any stop token). */
  outTokens: number[];
  /** Wall-clock decode latency (ms) on the node. */
  elapsedMs: number;
  /** The model the node reports (from --model or /status). */
  model: string;
  /** Diagnostic detail (present on no-node / tokenize-400 / decode error). */
  detail?: string;
  /** True when the reply was served from the GLOBAL ANSWER CACHE (no inference). */
  cached?: boolean;
  /** Which mycelial cache tier served a hit (fano7 / aeon66 / monster196884). */
  cacheTier?: string;
}

// --------------------------------------------------------------------------
// GLOBAL ANSWER CACHE (WS-3 cache-aware bridge path).
//
// The cache library + HTTP contract are owned by WS-1 (@a0n/knotgraph
// global-cache.ts) and WS-2 (edgework-app /api/v1/cache/*). @a0n/knotgraph is
// NOT a dependency of @a0n/gnosis, so the bridge does NOT import it; instead it
// REPLICATES the lib's two contract surfaces it needs, byte-for-byte:
//
//   1) canonicalQueryHash(tokens, qspecId) -- the fp48 key, computed via the
//      SAME `cacheFingerprint48HexSync` from @a0n/bitwise/cache-fp48 over the
//      SAME canonical byte layout (qspecId utf8, 0x00, then each token id as a
//      little-endian u32). MODEL-AGNOSTIC: model is provenance, NOT in the key.
//      One differing prompt token or qspec changes the bytes -> the key. This is
//      the exact query-level analogue of fanoMycelialTierKeyBytes the WS-1 lib uses.
//   2) the GlobalAnswerCacheEntry shape the cache PUT body expects.
//
// Signing is v1: the WS-1 default verifier (presenceOnlySignatureVerifier) is
// presence-based, so a non-empty DETERMINISTIC node sig over the entry's
// identity (queryHash + answerTokens + nodeId) suffices. A real ed25519/HMAC
// node key is a follow-on; this is noted as v1.
// --------------------------------------------------------------------------

/** The fixed quality-spec / gate id for skymesh mesh queries (cache key part). */
export const SKYMESH_QSPEC_ID = 'skymesh-query/v2';
/** Canonical-key token separator byte (mirrors global-cache.ts TOKEN_SEP). */
const CACHE_TOKEN_SEP = 0x00;

/**
 * Replicates @a0n/knotgraph global-cache.ts `canonicalQueryKeyBytes`:
 *   qspecId utf8 | 0x00 | token[0] LE u32 | token[1] ...
 * MODEL-AGNOSTIC (model is provenance, NOT in the key). Byte-for-byte identical
 * so the bridge's key matches the WS-1/WS-2 key.
 */
export function canonicalQueryKeyBytes(
  queryTokens: readonly number[],
  qspecId: string,
): Uint8Array {
  const enc = new TextEncoder();
  const qspecBytes = enc.encode(qspecId);
  const out = new Uint8Array(qspecBytes.length + 1 + queryTokens.length * 4);
  let offset = 0;
  out.set(qspecBytes, offset);
  offset += qspecBytes.length;
  out[offset++] = CACHE_TOKEN_SEP;
  const view = new DataView(out.buffer);
  for (let i = 0; i < queryTokens.length; i++) {
    view.setUint32(offset, queryTokens[i]! >>> 0, true);
    offset += 4;
  }
  return out;
}

/**
 * The KEY: canonical query hash. fp48 (12 lowercase hex chars) over
 * `(queryTokens, qspecId)` via the SAME `cacheFingerprint48HexSync` the WS-1 lib
 * uses. MODEL-AGNOSTIC. Deterministic: same query -> same hash. Mirrors
 * @a0n/knotgraph `canonicalQueryHash` (incl. the empty-tokens clinamen guard).
 */
export function canonicalQueryHash(
  queryTokens: readonly number[],
  qspecId: string,
): string {
  if (queryTokens.length === 0) {
    throw new RangeError('canonicalQueryHash: empty queryTokens (clinamen_swerve guard)');
  }
  return cacheFingerprint48HexSync(canonicalQueryKeyBytes(queryTokens, qspecId));
}

/** The trusted attestation carried by a stored answer (mirrors WS-1 shape). */
export interface GlobalAnswerCacheAttestation {
  gate: string;
  pass: boolean;
  token: number;
  logit: number;
  nodeId: string;
  admitted: boolean;
  sig: string;
  attesterDid?: string;
}

/** One global cache entry (mirrors @a0n/knotgraph GlobalAnswerCacheEntry). */
export interface GlobalAnswerCacheEntry {
  queryHash: string;
  queryTokens: number[];
  model: string;
  qspecId: string;
  answerTokens: number[];
  answerText: string;
  attestation: GlobalAnswerCacheAttestation;
  ts: number;
  /** Optional tier hint; the cache stores it. */
  tier?: string;
}

/**
 * Deterministic node signature (v1). The WS-1 default verifier is presence-based,
 * so a stable non-empty digest over the entry's identity (queryHash, model,
 * qspecId, answerTokens, nodeId) suffices for v1. fp48 over the canonical bytes
 * gives a deterministic, non-empty, tamper-coupled string. Real ed25519/HMAC node
 * keys are a follow-on (noted v1).
 */
export function signCacheEntryV1(
  queryHash: string,
  answerTokens: readonly number[],
  nodeId: string,
  model: string,
  qspecId: string,
): string {
  const enc = new TextEncoder();
  const head = enc.encode(`${queryHash} ${model} ${qspecId} ${nodeId} `);
  const bytes = new Uint8Array(head.length + answerTokens.length * 4);
  bytes.set(head, 0);
  const view = new DataView(bytes.buffer);
  for (let i = 0; i < answerTokens.length; i++) {
    view.setUint32(head.length + i * 4, answerTokens[i]! >>> 0, true);
  }
  return `sigv1:${cacheFingerprint48HexSync(bytes)}`;
}

/** A cache lookup response (WS-2 FIXED contract: GET .../cache/lookup). */
export interface CacheLookupResponse {
  hit: boolean;
  entry?: GlobalAnswerCacheEntry;
}

/**
 * On-hit verification before serving a cached answer (mirrors WS-1 `verifyEntry`
 * + `sameQueryIdentity`). Requires the trusted-attestation invariants (pass +
 * admitted + a non-empty sig + a replayable answer) AND the fp48 birthday-bound
 * identity check: the entry's own (queryTokens, qspecId) must re-hash to its
 * queryHash, must equal the REQUEST key, and the request's tokens must match
 * exactly. MODEL-AGNOSTIC: model is provenance, NOT identity — it is not
 * compared (any model's answer serves). The signature check is presence-based
 * (WS-1 default verifier; real crypto is a follow-on, noted v1).
 */
export function verifyCacheEntry(
  entry: GlobalAnswerCacheEntry,
  requestHash: string,
  requestTokens: readonly number[],
): boolean {
  const att = entry.attestation;
  if (!att || att.pass !== true || att.admitted !== true) return false;
  if (typeof att.sig !== 'string' || att.sig.length === 0) return false;
  if (!Array.isArray(entry.answerTokens) || entry.answerTokens.length === 0) return false;
  if (!Array.isArray(entry.queryTokens) || entry.queryTokens.length === 0) return false;
  // The stored entry must be internally consistent: its own identity re-hashes to
  // its queryHash, and that equals the request key.
  let rederived: string;
  try {
    rederived = canonicalQueryHash(entry.queryTokens, entry.qspecId);
  } catch {
    return false;
  }
  if (rederived !== entry.queryHash || entry.queryHash !== requestHash) return false;
  // Request-identity guard past the fp48 birthday bound: tokens + qspec exact
  // (model is provenance, not compared).
  if (entry.qspecId !== SKYMESH_QSPEC_ID) return false;
  if (entry.queryTokens.length !== requestTokens.length) return false;
  for (let i = 0; i < requestTokens.length; i++) {
    if (entry.queryTokens[i] !== requestTokens[i]) return false;
  }
  return true;
}

export interface BridgeConfig {
  /** WSS relay endpoint (the DO's /monitor/bridge upgrade). */
  readonly relay: string;
  /** Local aeon-monitor base, e.g. http://127.0.0.1:8722. No trailing /monitor. */
  readonly local: string;
  /** Shared secret presented to the relay on the WSS upgrade. */
  readonly token: string;
  /** Scenes to mirror; one local SSE subscription per entry. */
  readonly scenes: readonly string[];
  /** Status poll cadence (ms). */
  readonly statusPollMs: number;
  /** Backoff cap (ms) for relay reconnects. */
  readonly backoffCapMs: number;
  /** Local read retry cadence (ms) when the local monitor is down. */
  readonly localRetryMs: number;
  /** Mesh id this bridge publishes to (the relay routes per mesh id). */
  readonly meshId: string;
  /** Node id this bridge identifies as (frames carry it as metadata.nodeId). */
  readonly nodeId: string;
  /** When true, do NOT connect the hosted-relay uplink (DashRelay-only). */
  readonly noRelay: boolean;
  /** Optional DashRelay wss base (e.g. wss://dashrelay.com); '' = disabled. */
  readonly dashrelay: string;
  /** Path to the model the preflight prover loads (PARIS gate). `~` resolved. */
  readonly model: string;
  /**
   * Canonical model identity used in the GLOBAL ANSWER CACHE key (NOT the local
   * filesystem path -- two nodes with the same model at different paths MUST
   * share a cache key for the global cache to work cross-node). Defaults to the
   * model file's basename without extension; override with --model-id.
   */
  readonly modelId: string;
  /** Path to the paris-probe-rknot prover binary. `~` resolved. */
  readonly probeBin: string;
  /** Local fat-station inference node URL the bridge queries (decode-next path). */
  readonly nodeUrl: string;
  /**
   * GLOBAL ANSWER CACHE base URL (e.g. https://edgework.ai). '' = cache OFF: the
   * bridge never looks up or stores answers and every query runs inference. When
   * set, a mesh query first looks up the global cache and serves a verified hit,
   * else runs inference and (only if this node is admitted) stores the answer.
   */
  readonly cacheUrl: string;
}

const DEFAULT_RELAY = 'wss://skymesh.forkjoin.ai/monitor/bridge';
const DEFAULT_LOCAL = 'http://127.0.0.1:8722';
const DEFAULT_SCENES = 'skymesh,spectrum-waterfall,noaa-apt,image,mesh,residual,gate-telemetry,selftalk';
/** The local fat-station inference node the bridge queries (decode-next path). */
const DEFAULT_NODE_URL = 'http://127.0.0.1:8786';

/** Expand a leading `~` to the user's home directory; leave the rest as-is. */
export function resolveHome(p: string): string {
  if (p === '~') return os.homedir();
  if (p.startsWith('~/')) return path.join(os.homedir(), p.slice(2));
  return p;
}

/**
 * Resolve the default prover binary path: <gnosis repo root>/<DEFAULT_PROBE_REL>.
 * This .mts lives at <gnosis>/rtlsdr-mock-sim/skymesh-bridge.mts, so the gnosis
 * root is two levels up from the file's dir, and the repo root is one more up.
 */
function defaultProbeBin(): string {
  try {
    const here = new URL('.', import.meta.url).pathname; // .../rtlsdr-mock-sim/
    const repoRoot = path.resolve(here, '..', '..', '..'); // up to monorepo root
    return path.join(repoRoot, DEFAULT_PROBE_REL);
  } catch {
    return DEFAULT_PROBE_REL;
  }
}

/**
 * Canonical model identity for the GLOBAL cache key: the model file's basename
 * without its extension (e.g. /home/x/qwen2.5-0.5b-instruct.knot ->
 * "qwen2.5-0.5b-instruct"). Stable across machines/paths so nodes running the
 * SAME model share cache keys. Empty/odd input falls back to the raw string.
 */
export function canonicalModelId(modelPath: string): string {
  try {
    const base = path.basename(modelPath);
    // Strip ONLY a known model-file extension -- NOT any trailing dot-segment
    // (path.extname would mangle dotted names like "qwen2.5" -> "qwen2" or
    // "tinyllama-1.1b" -> "tinyllama-1"). Model weights are .knot/.gguf/etc.
    const ext = path.extname(base).toLowerCase();
    const KNOWN = new Set(['.knot', '.gguf', '.safetensors', '.bin', '.pt', '.pth', '.onnx']);
    const stem = KNOWN.has(ext) ? base.slice(0, -ext.length) : base;
    return stem.length > 0 ? stem : modelPath;
  } catch {
    return modelPath;
  }
}

/** Resolve the default node id (hostname), with a safe fallback. */
function defaultNodeId(): string {
  try {
    // os.hostname() via the env-agnostic require (the .mts runs under Node 22+).
    // Use a dynamic import-free path so the pure helpers stay test-friendly.
    return os.hostname() || 'skymesh-node';
  } catch {
    return 'skymesh-node';
  }
}

/** Build a config from argv + env (the runner / direct invocation path). */
export function configFromArgv(argv: readonly string[], env: NodeJS.ProcessEnv): BridgeConfig {
  return {
    relay: argValue(argv, '--relay', DEFAULT_RELAY),
    local: argValue(argv, '--local', DEFAULT_LOCAL).replace(/\/+$/, ''),
    token: argValue(argv, '--token', env.SKYMESH_BRIDGE_TOKEN ?? ''),
    scenes: argValue(argv, '--scenes', DEFAULT_SCENES)
      .split(',')
      .map((s) => s.trim())
      .filter((s) => s.length > 0),
    statusPollMs: Number.parseInt(argValue(argv, '--status-poll-ms', '1000'), 10),
    backoffCapMs: Number.parseInt(argValue(argv, '--backoff-cap-ms', '30000'), 10),
    localRetryMs: Number.parseInt(argValue(argv, '--local-retry-ms', '2000'), 10),
    meshId: argValue(argv, '--mesh-id', SKYMESH_GLOBAL_MESH_ID),
    nodeId: argValue(argv, '--node-id', defaultNodeId()),
    noRelay: argFlag(argv, '--no-relay'),
    dashrelay: argValue(argv, '--dashrelay', '').replace(/\/+$/, ''),
    model: resolveHome(argValue(argv, '--model', DEFAULT_PREFLIGHT_MODEL)),
    // Cache identity: --model-id wins; else the model file's basename (no ext).
    modelId: argValue(
      argv,
      '--model-id',
      canonicalModelId(resolveHome(argValue(argv, '--model', DEFAULT_PREFLIGHT_MODEL))),
    ),
    probeBin: resolveHome(argValue(argv, '--probe-bin', defaultProbeBin())),
    nodeUrl: argValue(argv, '--node-url', env.SKYMESH_NODE_URL ?? DEFAULT_NODE_URL).replace(/\/+$/, ''),
    // GLOBAL ANSWER CACHE: --cache-url turns it on; --no-cache forces it off even
    // if a URL is supplied (or via env SKYMESH_CACHE_URL). Default '' = cache OFF.
    cacheUrl: argFlag(argv, '--no-cache')
      ? ''
      : argValue(argv, '--cache-url', env.SKYMESH_CACHE_URL ?? '').replace(/\/+$/, ''),
  };
}

// --------------------------------------------------------------------------
// Logging (no emoji; connection state, frame counts, control ops).
// --------------------------------------------------------------------------

function log(...args: unknown[]): void {
  console.log('[skymesh-bridge]', ...args);
}
function logErr(...args: unknown[]): void {
  console.error('[skymesh-bridge]', ...args);
}

// --------------------------------------------------------------------------
// Tiny server-side SSE `data:` parser. Feeds raw stream chunks in, yields the
// JSON payload of each `data:` event out. Mirrors the local monitor's
// `data: <json>\n\n` framing (aeon-monitor.rs:5484). NOT the browser EventSource.
// --------------------------------------------------------------------------

export class SseDataParser {
  private buf = '';

  /**
   * Feed a raw text chunk; return the parsed payloads of every COMPLETE
   * `data:` event in this + buffered bytes. A `:`-comment heartbeat (`: ping`)
   * carries no `data:` line, so it is silently dropped. Partial events stay
   * buffered for the next chunk.
   */
  push(chunk: string): unknown[] {
    this.buf += chunk;
    const out: unknown[] = [];
    let sep: number;
    // Events are delimited by a blank line (\n\n). Handle \r\n\r\n too.
    while ((sep = this.nextDelimiter()) !== -1) {
      const rawEvent = this.buf.slice(0, sep);
      // Advance past the delimiter (2 for \n\n, 4 for \r\n\r\n).
      this.buf = this.buf.slice(sep + this.delimiterLen);
      const dataLines: string[] = [];
      for (const line of rawEvent.split(/\r?\n/)) {
        if (line.startsWith('data:')) {
          // Per the SSE spec a single leading space after the colon is stripped.
          dataLines.push(line.slice(5).replace(/^ /, ''));
        }
      }
      if (dataLines.length === 0) continue; // comment-only / heartbeat event
      const payload = dataLines.join('\n');
      if (payload.length === 0) continue;
      try {
        out.push(JSON.parse(payload));
      } catch {
        // A non-JSON data line is not a monitor envelope; skip it.
      }
    }
    return out;
  }

  private delimiterLen = 2;
  /** Index of the next event delimiter (\n\n or \r\n\r\n), or -1. */
  private nextDelimiter(): number {
    const lf = this.buf.indexOf('\n\n');
    const crlf = this.buf.indexOf('\r\n\r\n');
    if (crlf !== -1 && (lf === -1 || crlf < lf)) {
      this.delimiterLen = 4;
      return crlf;
    }
    if (lf !== -1) {
      this.delimiterLen = 2;
      return lf;
    }
    return -1;
  }
}

// --------------------------------------------------------------------------
// The bridge. One instance owns the relay socket lifecycle + the local
// subscriptions + the reconnect/backoff loop.
// --------------------------------------------------------------------------

/** Minimal shape of the WebSocket global we rely on (Node 22+ / ws-compatible).
 * The aeon-flow uplink sends binary Flow frames, so `send` accepts bytes too. */
export interface WsLike {
  send(data: string | Uint8Array): void;
  close(code?: number, reason?: string): void;
  addEventListener(type: 'open', cb: () => void): void;
  addEventListener(type: 'message', cb: (ev: { data: unknown }) => void): void;
  addEventListener(type: 'close', cb: () => void): void;
  addEventListener(type: 'error', cb: (ev?: unknown) => void): void;
  readyState: number;
}

/** Factory so tests can inject a mock socket; default uses the WebSocket global. */
export type WsFactory = (url: string, token: string) => WsLike;

const defaultWsFactory: WsFactory = (url, token) => {
  // The built-in WebSocket global (Node 22+) ignores a second options arg, so we
  // pass the token via ?token= (the relay supports the query fallback). When run
  // under a `ws`-style impl that accepts headers, also send the bearer header.
  const u = new URL(url);
  if (token) u.searchParams.set('token', token);
  const Ctor = (globalThis as { WebSocket?: unknown }).WebSocket as
    | (new (u: string, ...rest: unknown[]) => WsLike)
    | undefined;
  if (!Ctor) {
    throw new Error(
      'No global WebSocket. Run on Node 22+ (the repo runtime) or provide a WsFactory.',
    );
  }
  try {
    return new Ctor(u.toString(), undefined, {
      headers: token ? { Authorization: `Bearer ${token}` } : {},
    });
  } catch {
    // The undici WebSocket throws on the 3rd arg in some versions; fall back.
    return new Ctor(u.toString());
  }
};

const WS_OPEN = 1;

/**
 * Append the GLOBAL PARTYLINE routing params (`?mesh=&node=`) to a relay WS URL.
 * The relay routes per mesh id; the node id attributes each frame to a node.
 * Preserves any existing query (e.g. an operator-supplied path).
 */
export function relayUrlWithMesh(relay: string, meshId: string, nodeId: string): string {
  try {
    const u = new URL(relay);
    u.searchParams.set('mesh', meshId);
    u.searchParams.set('node', nodeId);
    return u.toString();
  } catch {
    // Not a parseable absolute URL (shouldn't happen for a wss relay); fall back
    // to a manual append so the bridge still carries the params.
    const sep = relay.includes('?') ? '&' : '?';
    return `${relay}${sep}mesh=${encodeURIComponent(meshId)}&node=${encodeURIComponent(nodeId)}`;
  }
}

/**
 * Build the DashRelay aeon-flow room URL: `${wssBase}/.aeon/flow?room=${meshId}`.
 * This is the LIVE endpoint (handshake 101 at dashrelay.com); it speaks the aeon
 * Flow wire (binary FlowCodec frames) after the same JSON join the relay uses.
 */
export function dashrelayFlowUrl(wssBase: string, meshId: string): string {
  const base = wssBase.replace(/\/+$/, '');
  return `${base}/.aeon/flow?room=${encodeURIComponent(meshId)}`;
}

export class SkymeshBridge {
  private readonly cfg: BridgeConfig;
  private readonly wsFactory: WsFactory;
  private readonly fetchImpl: typeof fetch;
  private readonly probeRunner: ProbeRunner;

  private ws: WsLike | null = null;
  private connected = false;
  private stopped = false;

  /**
   * Set true once a PREFLIGHT PARIS gate has PASSED on this bridge. Carried in
   * the hello and every status sent up (additive) so the relay/app can mark this
   * node admitted. A node proves itself with REAL inference BEFORE joining global.
   */
  private admitted = false;

  private statusTimer: ReturnType<typeof setInterval> | null = null;
  private reconnectTimer: ReturnType<typeof setTimeout> | null = null;
  private backoffMs = 1000;
  private lastStatusJson = '';

  // ---- DashRelay uplink (optional, additive second fan-out) ----
  private dashWs: WsLike | null = null;
  private dashConnected = false;
  private dashJoined = false;
  private dashReconnectTimer: ReturnType<typeof setTimeout> | null = null;
  private dashBackoffMs = 1000;
  private dashFramesUp = 0;
  // JS-only FlowCodec (no WASM): encodes each monitor envelope into the 10-byte
  // self-describing Flow frame on the MonitorFlow stream before broadcast.
  private readonly flowCodec: FlowCodec = FlowCodec.createSync();

  // Per-scene local SSE subscriptions: an AbortController to tear each down, and
  // a retry timer used when the local monitor is unreachable.
  private readonly localSubs = new Map<string, AbortController>();
  private readonly localRetryTimers = new Map<string, ReturnType<typeof setTimeout>>();
  // A master controller: stop() aborts it to kill EVERY in-flight fetch (status
  // poll, scenes, control, SSE) at once, so no handle survives teardown. This is
  // what lets the in-process test runner drain its event loop after stop().
  private lifeCtl = new AbortController();

  // Counters for clean logging.
  private framesUp = 0;
  private controlsApplied = 0;

  /**
   * The fetch the mesh-query runner uses. Defaults to a wrapper over the
   * bridge's fetchImpl; tests inject a mock fat-station here (the REAL decode
   * still runs against whatever this returns -- the reply is never mocked). When
   * an explicit queryFetch is not supplied, it adapts the bridge fetchImpl.
   */
  private readonly queryFetch: QueryFetch;

  constructor(
    cfg: BridgeConfig,
    opts?: {
      wsFactory?: WsFactory;
      fetchImpl?: typeof fetch;
      probeRunner?: ProbeRunner;
      queryFetch?: QueryFetch;
    },
  ) {
    this.cfg = cfg;
    this.wsFactory = opts?.wsFactory ?? defaultWsFactory;
    this.fetchImpl = opts?.fetchImpl ?? fetch;
    this.probeRunner = opts?.probeRunner ?? defaultProbeRunner;
    const f = this.fetchImpl;
    this.queryFetch =
      opts?.queryFetch ??
      ((url, init) =>
        f(url, init as RequestInit) as unknown as Promise<{
          ok: boolean;
          status: number;
          text: () => Promise<string>;
        }>);
  }

  /** Begin the bridge: open the relay and keep it open until stop(). */
  start(): void {
    this.stopped = false;
    if (this.lifeCtl.signal.aborted) this.lifeCtl = new AbortController();
    log(`mesh ${this.cfg.meshId}  node ${this.cfg.nodeId}`);
    log(`mirroring scenes: ${this.cfg.scenes.join(', ')}`);
    if (this.cfg.noRelay) {
      log('hosted relay disabled (--no-relay); DashRelay-only');
    } else {
      log(`local ${this.cfg.local}  ->  relay ${this.cfg.relay}`);
      this.connect();
    }
    if (this.cfg.dashrelay) {
      log(`local ${this.cfg.local}  ->  dashrelay ${this.cfg.dashrelay} room=${this.cfg.meshId}`);
      this.connectDash();
    }
    if (this.cfg.noRelay && !this.cfg.dashrelay) {
      logErr('no uplink: --no-relay set without --dashrelay; nothing to publish to');
    }
  }

  /** Tear everything down (tests + clean shutdown). Idempotent. */
  stop(): void {
    this.stopped = true;
    if (this.reconnectTimer) clearTimeout(this.reconnectTimer);
    this.reconnectTimer = null;
    if (this.dashReconnectTimer) clearTimeout(this.dashReconnectTimer);
    this.dashReconnectTimer = null;
    this.teardownRelaySideEffects();
    // Abort every in-flight fetch (SSE + status/scenes/control). Per-scene
    // controllers are also aborted in teardownRelaySideEffects; the master
    // controller covers requests not tied to a scene.
    try {
      this.lifeCtl.abort();
    } catch {
      /* ignore */
    }
    try {
      this.ws?.close(1000, 'bridge stop');
    } catch {
      /* ignore */
    }
    this.ws = null;
    try {
      this.dashWs?.close(1000, 'bridge stop');
    } catch {
      /* ignore */
    }
    this.dashWs = null;
    this.dashConnected = false;
    this.dashJoined = false;
    this.stopDashLocalPump();
  }

  // ---- relay socket lifecycle ----

  private connect(): void {
    if (this.stopped) return;
    let ws: WsLike;
    try {
      // Carry the GLOBAL PARTYLINE routing on the upgrade (?mesh=&node=); the
      // first WS message is a {type:"hello", nodeId} (see onRelayOpen).
      const relayUrl = relayUrlWithMesh(this.cfg.relay, this.cfg.meshId, this.cfg.nodeId);
      ws = this.wsFactory(relayUrl, this.cfg.token);
    } catch (e) {
      logErr(`relay connect failed: ${errMsg(e)}; backing off`);
      this.scheduleReconnect();
      return;
    }
    this.ws = ws;

    ws.addEventListener('open', () => {
      if (this.ws !== ws) return; // a stale socket fired late
      this.connected = true;
      this.backoffMs = 1000; // reset backoff on a clean open
      log('relay connected (open)');
      void this.onRelayOpen(ws);
    });

    ws.addEventListener('message', (ev) => {
      if (this.ws !== ws) return;
      void this.onRelayMessage(ws, ev.data);
    });

    ws.addEventListener('close', () => {
      if (this.ws !== ws) return;
      const wasConnected = this.connected;
      this.connected = false;
      this.teardownRelaySideEffects();
      if (this.stopped) return;
      if (wasConnected) log('relay closed; reconnecting with backoff');
      this.scheduleReconnect();
    });

    ws.addEventListener('error', (e) => {
      if (this.ws !== ws) return;
      logErr(`relay socket error: ${errMsg(e)}`);
      try {
        ws.close();
      } catch {
        /* the close handler drives the reconnect */
      }
    });
  }

  private scheduleReconnect(): void {
    if (this.stopped || this.reconnectTimer) return;
    const jitter = Math.floor(Math.random() * 250);
    const delay = Math.min(this.backoffMs, this.cfg.backoffCapMs) + jitter;
    log(`reconnecting in ${delay}ms`);
    this.reconnectTimer = setTimeout(() => {
      this.reconnectTimer = null;
      // Exponential growth, capped.
      this.backoffMs = Math.min(this.backoffMs * 2, this.cfg.backoffCapMs);
      this.connect();
    }, delay);
    // NOTE: do NOT unref() this timer. On a relay drop, teardownRelaySideEffects()
    // closes the SSE streams + status interval, so the pending reconnect timer is
    // the only thing holding the event loop open -- unref'ing it lets Node exit
    // mid-backoff and the bridge silently dies (observed when the relay was
    // redeployed). stop() clears this timer explicitly, so a clean shutdown does
    // not need the unref.
  }

  /** On a fresh open: send hello (node identity), push scenes + status, start
   * polling, (re)subscribe local. */
  private async onRelayOpen(ws: WsLike): Promise<void> {
    // First frame of identity for the GLOBAL PARTYLINE: tells the relay which
    // node this socket is so returned frames can carry metadata.nodeId. The
    // `admitted` flag (additive) carries the PARIS-gate verdict across reconnects.
    this.sendUp(ws, { type: 'hello', nodeId: this.cfg.nodeId, admitted: this.admitted });
    await this.pushScenes(ws);
    await this.pushStatus(ws, true);
    this.startStatusPolling();
    // (Re)subscribe every mirrored scene to the local SSE.
    for (const scene of this.cfg.scenes) this.subscribeLocalScene(ws, scene);
  }

  /** A DOWN message from the relay; `control`, `preflight`, `query` are acted on. */
  private async onRelayMessage(ws: WsLike, raw: unknown): Promise<void> {
    let msg: {
      type?: string;
      id?: string;
      op?: string;
      args?: unknown;
      prompt?: unknown;
      tokens?: unknown;
      maxTokens?: unknown;
    };
    try {
      const text =
        typeof raw === 'string'
          ? raw
          : raw instanceof ArrayBuffer
            ? new TextDecoder().decode(raw)
            : Buffer.isBuffer(raw)
              ? raw.toString('utf8')
              : String(raw);
      msg = JSON.parse(text);
    } catch {
      return; // ignore non-JSON
    }
    if (msg.type === 'control') {
      await this.applyControl(ws, msg.id, msg.args);
      return;
    }
    if (msg.type === 'preflight') {
      await this.runPreflight(ws, msg.id);
      return;
    }
    if (msg.type === 'query') {
      await this.handleQuery(ws, msg.id, {
        prompt: typeof msg.prompt === 'string' ? msg.prompt : undefined,
        tokens: Array.isArray(msg.tokens)
          ? msg.tokens.filter((n): n is number => typeof n === 'number')
          : undefined,
        maxTokens: typeof msg.maxTokens === 'number' ? msg.maxTokens : undefined,
      });
      return;
    }
  }

  // ---- DOWN: query (run the REAL decode-next mesh query, reply UP) ----

  /**
   * Run the mesh inference query against the local fat-station (cfg.nodeUrl) via
   * the VERIFIED decode-next path and reply {type:"query-result",...} UP. Never
   * throws -- on no node / tokenize 400 / decode error the result carries a
   * `detail` explaining the failure. The decode is REAL (never a mocked reply).
   */
  private async handleQuery(ws: WsLike, id: string | undefined, input: MeshQueryInput): Promise<void> {
    const reqId = typeof id === 'string' && id.length > 0 ? id : crypto.randomUUID();
    log(
      `query ${reqId}: decode-next against ${this.cfg.nodeUrl} ` +
        `(${input.tokens ? `${input.tokens.length} raw tokens` : `prompt ${JSON.stringify(input.prompt ?? '')}`}, ` +
        `maxTokens ${input.maxTokens ?? QUERY_DEFAULT_MAX_TOKENS})`,
    );
    const outcome = await runMeshQuery(
      this.cfg.nodeUrl,
      input,
      this.queryFetch,
      // Use the CANONICAL model id (basename, no ext) as the query + cache model
      // identity -- NOT the local filesystem path -- so the GLOBAL cache key is
      // shared across nodes running the same model. Falls back to the node's
      // /status model when unset.
      this.cfg.modelId || undefined,
      // GLOBAL ANSWER CACHE: pass the cache seam only when a cache URL is set.
      this.cfg.cacheUrl ? this.cacheHooks() : undefined,
    );
    if (outcome.detail) {
      log(`query ${reqId}: detail -- ${outcome.detail} (elapsed=${outcome.elapsedMs}ms)`);
    } else if (outcome.cached) {
      log(
        `query ${reqId}: SERVED FROM GLOBAL CACHE` +
          `${outcome.cacheTier ? ` (tier ${outcome.cacheTier})` : ''} ` +
          `${outcome.outTokens.length} tok in ${outcome.elapsedMs}ms ` +
          `reply=${JSON.stringify(outcome.reply.slice(0, 80))}`,
      );
    } else {
      log(
        `query ${reqId}: ${outcome.outTokens.length} tok in ${outcome.elapsedMs}ms ` +
          `reply=${JSON.stringify(outcome.reply.slice(0, 80))}`,
      );
    }
    const result: QueryResult = {
      type: 'query-result',
      id: reqId,
      reply: outcome.reply,
      promptTokens: outcome.promptTokens,
      outTokens: outcome.outTokens,
      elapsedMs: outcome.elapsedMs,
      model: outcome.model,
      ...(outcome.detail ? { detail: outcome.detail } : {}),
      // Surface the cache flag whenever a cache is configured (true on a hit,
      // false on a fresh inference); omitted entirely when the cache is OFF.
      ...(this.cfg.cacheUrl ? { cached: outcome.cached === true } : {}),
      ...(outcome.cacheTier ? { cacheTier: outcome.cacheTier } : {}),
    };
    this.sendUp(ws, result as unknown as Record<string, unknown>);
  }

  // ---- GLOBAL ANSWER CACHE seam (WS-3) ----

  /**
   * Build the cache hooks runMeshQuery calls. `lookup` GETs the WS-2 cache and
   * serves only a VERIFIED hit (pass + admitted + present sig + identity match);
   * `store` POSTs a signed entry ONLY when THIS bridge is admitted. Both are
   * graceful: any cache-side error returns null / no-ops so the query never
   * breaks and never throws. Quality is preserved -- a hit is verified before it
   * is served, and only admitted+pass+signed entries are stored.
   */
  private cacheHooks(): MeshQueryCacheHooks {
    return {
      lookup: (promptTokens, model) => this.cacheLookup(promptTokens, model),
      store: (promptTokens, outTokens, reply, model) =>
        this.cacheStore(promptTokens, outTokens, reply, model),
    };
  }

  /**
   * GET {cacheUrl}/api/v1/cache/lookup?q=&model=&qspec=&tokens=<csv> (WS-2 FIXED
   * contract). On {hit:true, entry} whose attestation verifies (pass + admitted +
   * non-empty sig) AND whose identity re-hashes to the request key, return the
   * cached answer. Anything else (miss, unreachable, malformed, unverified) ->
   * null (the caller runs inference). NEVER throws.
   */
  private async cacheLookup(
    promptTokens: number[],
    model: string,
  ): Promise<MeshQueryCacheHit | null> {
    let queryHash: string;
    try {
      queryHash = canonicalQueryHash(promptTokens, SKYMESH_QSPEC_ID);
    } catch {
      return null; // empty tokens (clinamen guard) -- nothing to key
    }
    const url =
      `${this.cfg.cacheUrl}/api/v1/cache/lookup` +
      `?q=${encodeURIComponent(queryHash)}` +
      `&model=${encodeURIComponent(model)}` +
      `&qspec=${encodeURIComponent(SKYMESH_QSPEC_ID)}` +
      `&tokens=${encodeURIComponent(promptTokens.join(','))}`;
    let body: CacheLookupResponse | null = null;
    try {
      const res = await this.queryFetch(url, { method: 'GET' });
      if (!res.ok) return null;
      body = JSON.parse(await res.text()) as CacheLookupResponse;
    } catch (e) {
      log(`cache lookup unreachable (${errMsg(e)}); running inference`);
      return null;
    }
    if (!body || body.hit !== true || !body.entry) return null;
    const entry = body.entry;
    // VERIFY before serving: trusted-attestation invariants + fp48 birthday-bound
    // identity (re-hash + exact token/qspec match against the request; model is
    // provenance, not compared).
    if (!verifyCacheEntry(entry, queryHash, promptTokens)) {
      log(`cache hit for ${queryHash} FAILED verification; running inference`);
      return null;
    }
    return {
      outTokens: entry.answerTokens,
      reply: typeof entry.answerText === 'string' ? entry.answerText : '',
      ...(typeof entry.tier === 'string' ? { tier: entry.tier } : {}),
    };
  }

  /**
   * POST {cacheUrl}/api/v1/cache/put with a signed GlobalAnswerCacheEntry, ONLY
   * when THIS bridge is admitted (quality gate: an un-admitted node never seeds
   * the global cache). The entry carries the PARIS-gate attestation (pass +
   * admitted + a deterministic v1 sig). Best-effort: any failure is logged and
   * ignored. NEVER throws.
   */
  private async cacheStore(
    promptTokens: number[],
    outTokens: number[],
    reply: string,
    model: string,
  ): Promise<void> {
    // Only admitted nodes may seed the cache. (validateEntryForPut on the WS-1
    // side also rejects admitted!==true, but we refuse to even POST here.)
    if (!this.admitted) {
      log('cache store skipped: node not admitted (only admitted nodes seed the cache)');
      return;
    }
    if (outTokens.length === 0) return;
    let queryHash: string;
    try {
      queryHash = canonicalQueryHash(promptTokens, SKYMESH_QSPEC_ID);
    } catch {
      return;
    }
    const entry: GlobalAnswerCacheEntry = {
      queryHash,
      queryTokens: promptTokens,
      model,
      qspecId: SKYMESH_QSPEC_ID,
      answerTokens: outTokens,
      answerText: reply,
      attestation: {
        gate: PREFLIGHT_TEST_ID,
        pass: this.admitted,
        token: outTokens[0] ?? 0,
        logit: 0,
        nodeId: this.cfg.nodeId,
        admitted: this.admitted,
        sig: signCacheEntryV1(queryHash, outTokens, this.cfg.nodeId, model, SKYMESH_QSPEC_ID),
        attesterDid: this.cfg.nodeId,
      },
      ts: Date.now(),
    };
    try {
      const res = await this.queryFetch(`${this.cfg.cacheUrl}/api/v1/cache/put`, {
        method: 'POST',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify(entry),
      });
      if (res.ok) {
        log(`cache store: PUT ${queryHash} ok`);
      } else {
        log(`cache store: PUT ${queryHash} HTTP ${res.status} (ignored)`);
      }
    } catch (e) {
      log(`cache store: PUT ${queryHash} failed (${errMsg(e)}; ignored)`);
    }
  }

  // ---- DOWN: preflight (run the PARIS gate, reply preflight-result UP) ----

  /**
   * Run the PREFLIGHT PARIS gate: spawn the VERIFIED prover (REAL inference), parse
   * the verdict, reply {type:"preflight-result",...} UP, and on PASS set the
   * internal `admitted` flag (so subsequent hello/status carry it). Never throws --
   * a spawn/timeout failure replies {pass:false, detail:"probe error: ..."}.
   */
  private async runPreflight(ws: WsLike, id: string | undefined): Promise<void> {
    const reqId = typeof id === 'string' && id.length > 0 ? id : PREFLIGHT_TEST_ID;
    log(`preflight ${reqId}: running the Paris quality gate (model ${this.cfg.model})`);
    const started = Date.now();
    let result: PreflightResult;
    try {
      const run = await this.probeRunner(this.cfg.probeBin, this.cfg.model, 25_000);
      const parsed = parseProbeOutput(run);
      result = {
        type: 'preflight-result',
        id: reqId,
        pass: parsed.pass,
        token: parsed.token,
        text: PREFLIGHT_EXPECTED_TEXT,
        logit: parsed.logit,
        elapsedMs: Date.now() - started,
        model: this.cfg.model,
        ...(parsed.detail ? { detail: parsed.detail } : {}),
      };
    } catch (e) {
      // The runner should never reject, but stay fail-soft regardless.
      result = {
        type: 'preflight-result',
        id: reqId,
        pass: false,
        token: 0,
        text: PREFLIGHT_EXPECTED_TEXT,
        logit: 0,
        elapsedMs: Date.now() - started,
        model: this.cfg.model,
        detail: `probe error: ${errMsg(e)}`,
      };
    }

    if (result.pass) {
      this.admitted = true;
      log(
        `preflight ${reqId}: PASS token=${result.token} logit=${result.logit} ` +
          `elapsed=${result.elapsedMs}ms -- node ADMITTED`,
      );
      // Re-announce admitted status up so the relay flips this node immediately
      // (additive; the next status poll would also carry it).
      void this.pushStatus(ws, true);
    } else {
      log(`preflight ${reqId}: FAIL ${result.detail ?? ''} (elapsed=${result.elapsedMs}ms)`);
    }
    this.sendUp(ws, result as unknown as Record<string, unknown>);
  }

  // ---- UP: scenes ----

  private async pushScenes(ws: WsLike): Promise<void> {
    const scenes = await this.localGetJson('/monitor/scenes');
    if (scenes == null) {
      logErr('local /monitor/scenes unavailable at open; will refresh on next poll');
      return;
    }
    this.sendUp(ws, { type: 'scenes', scenes });
    log('sent scenes catalog up');
  }

  // ---- UP: status (open + ~1s poll on change) ----

  private startStatusPolling(): void {
    this.stopStatusPolling();
    this.statusTimer = setInterval(() => {
      if (this.ws && this.connected) void this.pushStatus(this.ws, false);
    }, this.cfg.statusPollMs);
    if (typeof this.statusTimer.unref === 'function') this.statusTimer.unref();
  }

  private stopStatusPolling(): void {
    if (this.statusTimer) clearInterval(this.statusTimer);
    this.statusTimer = null;
  }

  /** Poll local status; send UP on change (or always on `force`). */
  private async pushStatus(ws: WsLike, force: boolean): Promise<void> {
    const status = await this.localGetJson('/monitor/status');
    if (status == null) return; // local down; keep relay alive, retry next tick
    // Fold the PARIS-gate verdict into the change key so flipping `admitted`
    // (after a preflight PASS) is always pushed even if the local status is
    // byte-stable.
    const json = JSON.stringify({ s: status, a: this.admitted });
    if (!force && json === this.lastStatusJson) return; // unchanged: skip
    this.lastStatusJson = json;
    // `admitted` rides alongside the status (additive; the relay reads it off
    // the message, leaving the local-monitor status payload byte-compatible).
    this.sendUp(ws, { type: 'status', status, admitted: this.admitted });
    log('sent status up' + (force ? ' (initial)' : ' (changed)'));
  }

  // ---- UP: frames (one local SSE subscription per mirrored scene) ----

  private subscribeLocalScene(ws: WsLike, scene: string): void {
    if (this.stopped || this.ws !== ws) return;
    // Drop any prior subscription for this scene first.
    this.unsubscribeLocalScene(scene);
    const controller = new AbortController();
    this.localSubs.set(scene, controller);
    void this.runLocalScene(ws, scene, controller);
  }

  private unsubscribeLocalScene(scene: string): void {
    const prev = this.localSubs.get(scene);
    if (prev) {
      try {
        prev.abort();
      } catch {
        /* ignore */
      }
      this.localSubs.delete(scene);
    }
    const t = this.localRetryTimers.get(scene);
    if (t) {
      clearTimeout(t);
      this.localRetryTimers.delete(scene);
    }
  }

  /** Open the local SSE for one scene and forward each envelope UP. */
  private async runLocalScene(
    ws: WsLike,
    scene: string,
    controller: AbortController,
  ): Promise<void> {
    if (this.stopped) return;
    const url = `${this.cfg.local}/monitor/stream?scene=${encodeURIComponent(scene)}`;
    // Abort when EITHER the per-scene controller (re-subscribe) or the master
    // controller (stop) fires.
    const signal = AbortSignal.any([controller.signal, this.lifeCtl.signal]);
    try {
      const res = await this.fetchImpl(url, {
        headers: { accept: 'text/event-stream' },
        signal,
      });
      if (!res.ok || !res.body) {
        throw new Error(`local stream ${scene} HTTP ${res.status}`);
      }
      log(`local SSE subscribed: scene=${scene}`);
      const parser = new SseDataParser();
      const reader = res.body.getReader();
      const decoder = new TextDecoder();
      // eslint-disable-next-line no-constant-condition
      while (true) {
        const { value, done } = await reader.read();
        if (done) break;
        if (signal.aborted || this.stopped) break;
        const chunk = decoder.decode(value, { stream: true });
        for (const envelope of parser.push(chunk)) {
          if (this.ws !== ws || !this.connected) return; // relay gone; stop pushing
          this.forwardFrame(ws, envelope);
        }
      }
      // Stream ended cleanly (local monitor closed it); retry unless torn down.
      if (!signal.aborted) this.retryLocalScene(ws, scene);
    } catch (e) {
      if (signal.aborted || this.stopped) return; // intentional teardown
      logErr(`local SSE scene=${scene} error: ${errMsg(e)}; retrying in ${this.cfg.localRetryMs}ms`);
      this.retryLocalScene(ws, scene);
    }
  }

  /** Schedule a local re-subscription for one scene (local monitor down/looping). */
  private retryLocalScene(ws: WsLike, scene: string): void {
    if (this.stopped || this.ws !== ws) return;
    if (this.localRetryTimers.has(scene)) return;
    const t = setTimeout(() => {
      this.localRetryTimers.delete(scene);
      if (this.stopped || this.ws !== ws || !this.connected) return;
      this.subscribeLocalScene(ws, scene);
    }, this.cfg.localRetryMs);
    this.localRetryTimers.set(scene, t);
    if (typeof t.unref === 'function') t.unref();
  }

  private forwardFrame(ws: WsLike, envelope: unknown): void {
    // The relay requires envelope.scene to be a string; the local monitor always
    // sets it. Skip a malformed frame rather than poison the relay cache.
    const scene = (envelope as { scene?: unknown } | null)?.scene;
    if (typeof scene !== 'string') return;
    if (ws.readyState === WS_OPEN && this.connected) {
      this.sendUp(ws, { type: 'frame', envelope });
      this.framesUp++;
      if (this.framesUp % 50 === 0) log(`frames forwarded up: ${this.framesUp}`);
    }
    // ADDITIVE: also fan the frame out to the DashRelay aeon-flow room (as a
    // MonitorFlow Flow frame).
    this.publishDash({ type: 'frame', envelope });
  }

  // ---- DashRelay uplink (optional, additive) ----

  /** Open the DashRelay aeon-flow room socket and keep it open until stop(). */
  private connectDash(): void {
    if (this.stopped || !this.cfg.dashrelay) return;
    let ws: WsLike;
    try {
      const url = dashrelayFlowUrl(this.cfg.dashrelay, this.cfg.meshId);
      // No bridge token on the DashRelay path; the room is the routing key.
      ws = this.wsFactory(url, '');
    } catch (e) {
      logErr(`dashrelay connect failed: ${errMsg(e)}; backing off`);
      this.scheduleDashReconnect();
      return;
    }
    this.dashWs = ws;

    ws.addEventListener('open', () => {
      if (this.dashWs !== ws) return;
      this.dashConnected = true;
      this.dashJoined = false;
      this.dashBackoffMs = 1000;
      log('dashrelay connected (open)');
      // aeon-flow join: the SAME JSON JOIN the relay uses (room == meshId,
      // mode opaque-relay so binary Flow frames relay verbatim, clientId ==
      // nodeId). The local pump (DashRelay-only mode) starts on the
      // {type:"joined"} ack so frames publish only after the room accepts us
      // (onDashMessage). The DO follows joined with a {type:"flow_ready"}.
      this.sendDashRaw(ws, {
        type: 'join',
        room: this.cfg.meshId,
        mode: 'opaque-relay',
        clientId: this.cfg.nodeId,
      });
    });

    ws.addEventListener('message', (ev) => {
      if (this.dashWs !== ws) return;
      this.onDashMessage(ev.data);
    });

    ws.addEventListener('close', () => {
      if (this.dashWs !== ws) return;
      const was = this.dashConnected;
      this.dashConnected = false;
      this.dashJoined = false;
      if (this.cfg.noRelay) this.stopDashLocalPump();
      if (this.stopped) return;
      if (was) log('dashrelay closed; reconnecting with backoff');
      this.scheduleDashReconnect();
    });

    ws.addEventListener('error', (e) => {
      if (this.dashWs !== ws) return;
      logErr(`dashrelay socket error: ${errMsg(e)}`);
      try {
        ws.close();
      } catch {
        /* the close handler drives the reconnect */
      }
    });
  }

  private scheduleDashReconnect(): void {
    if (this.stopped || this.dashReconnectTimer || !this.cfg.dashrelay) return;
    const jitter = Math.floor(Math.random() * 250);
    const delay = Math.min(this.dashBackoffMs, this.cfg.backoffCapMs) + jitter;
    log(`dashrelay reconnecting in ${delay}ms`);
    this.dashReconnectTimer = setTimeout(() => {
      this.dashReconnectTimer = null;
      this.dashBackoffMs = Math.min(this.dashBackoffMs * 2, this.cfg.backoffCapMs);
      this.connectDash();
    }, delay);
    // Do NOT unref() -- same trap as the relay reconnect timer: on a DashRelay
    // drop this pending timer can be the only thing holding the event loop open,
    // and unref'ing it lets a DashRelay-only bridge silently exit mid-backoff.
    // stop() clears this timer explicitly, so clean shutdown does not need it.
  }

  /** Note the {type:"joined"} ack so we publish only after the room accepts us. */
  private onDashMessage(raw: unknown): void {
    let msg: { type?: string };
    try {
      const text =
        typeof raw === 'string'
          ? raw
          : raw instanceof ArrayBuffer
            ? new TextDecoder().decode(raw)
            : Buffer.isBuffer(raw)
              ? raw.toString('utf8')
              : String(raw);
      msg = JSON.parse(text);
    } catch {
      return;
    }
    if (msg.type === 'joined') {
      this.dashJoined = true;
      log(`dashrelay joined room=${this.cfg.meshId}`);
      // If the hosted relay is disabled, the DashRelay socket drives the local
      // pump. Start it only AFTER joined so the initial scenes/status land
      // (publishDash no-ops before join).
      if (this.cfg.noRelay) this.startDashLocalPump();
    }
  }

  /** Send a JSON control frame (the join handshake) to the DashRelay socket. */
  private sendDashRaw(ws: WsLike, msg: Record<string, unknown>): void {
    if (ws.readyState !== WS_OPEN) return;
    try {
      ws.send(JSON.stringify(msg));
    } catch (e) {
      logErr(`dashrelay send failed (${String(msg.type)}): ${errMsg(e)}`);
    }
  }

  /**
   * Publish a monitor payload ({type:"frame"|"status"|"scenes", ...}) to the
   * DashRelay aeon-flow room. Only `frame` payloads carry a monitor envelope;
   * each is encoded as a single MonitorFlow Flow frame
   * (`FlowCodec.encode(encodeMonitorFlowFrame(envelope))`) and broadcast as
   * binary. status/scenes carry no MonitorFlowFrame and are skipped on this wire
   * (the App consumes only `frame` envelopes from the DashRelay source). No-op
   * when the DashRelay uplink is disabled or not yet joined. Never throws.
   */
  private publishDash(payload: Record<string, unknown>): void {
    const ws = this.dashWs;
    if (!ws || !this.dashConnected || !this.dashJoined) return;
    if (payload.type !== 'frame') return;
    const frame = monitorFlowFrameFromEnvelope(payload.envelope);
    if (!frame) return;
    let bytes: Uint8Array;
    try {
      bytes = this.flowCodec.encode(encodeMonitorFlowFrame(frame));
    } catch (e) {
      logErr(`dashrelay flow encode failed: ${errMsg(e)}`);
      return;
    }
    try {
      ws.send(bytes);
    } catch (e) {
      logErr(`dashrelay flow send failed: ${errMsg(e)}`);
      return;
    }
    this.dashFramesUp++;
    if (this.dashFramesUp % 50 === 0) log(`dashrelay frames forwarded: ${this.dashFramesUp}`);
  }

  // ---- DashRelay-only local pump (drives local SSE when --no-relay) ----
  // The relay-coupled subscribeLocalScene path is keyed on the hosted-relay
  // socket; with --no-relay there is no such socket, so this independent pump
  // subscribes the local SSE + polls status and fans everything to publishDash.

  private dashPumpCtls = new Map<string, AbortController>();
  private dashPumpRetryTimers = new Map<string, ReturnType<typeof setTimeout>>();
  private dashStatusTimer: ReturnType<typeof setInterval> | null = null;
  private dashLastStatusJson = '';

  private startDashLocalPump(): void {
    void this.pushDashScenes();
    void this.pushDashStatus(true);
    this.stopDashStatusPolling();
    this.dashStatusTimer = setInterval(() => {
      if (this.dashConnected) void this.pushDashStatus(false);
    }, this.cfg.statusPollMs);
    if (typeof this.dashStatusTimer.unref === 'function') this.dashStatusTimer.unref();
    for (const scene of this.cfg.scenes) this.subscribeDashScene(scene);
  }

  private stopDashLocalPump(): void {
    this.stopDashStatusPolling();
    this.dashLastStatusJson = '';
    for (const [scene, ctl] of this.dashPumpCtls) {
      try {
        ctl.abort();
      } catch {
        /* ignore */
      }
      this.dashPumpCtls.delete(scene);
    }
    for (const [scene, t] of this.dashPumpRetryTimers) {
      clearTimeout(t);
      this.dashPumpRetryTimers.delete(scene);
    }
  }

  private stopDashStatusPolling(): void {
    if (this.dashStatusTimer) clearInterval(this.dashStatusTimer);
    this.dashStatusTimer = null;
  }

  private async pushDashScenes(): Promise<void> {
    const scenes = await this.localGetJson('/monitor/scenes');
    if (scenes == null) return;
    this.publishDash({ type: 'scenes', scenes });
  }

  private async pushDashStatus(force: boolean): Promise<void> {
    const status = await this.localGetJson('/monitor/status');
    if (status == null) return;
    const json = JSON.stringify(status);
    if (!force && json === this.dashLastStatusJson) return;
    this.dashLastStatusJson = json;
    this.publishDash({ type: 'status', status });
  }

  private subscribeDashScene(scene: string): void {
    if (this.stopped) return;
    const prev = this.dashPumpCtls.get(scene);
    if (prev) {
      try {
        prev.abort();
      } catch {
        /* ignore */
      }
    }
    const controller = new AbortController();
    this.dashPumpCtls.set(scene, controller);
    void this.runDashScene(scene, controller);
  }

  private async runDashScene(scene: string, controller: AbortController): Promise<void> {
    if (this.stopped) return;
    const url = `${this.cfg.local}/monitor/stream?scene=${encodeURIComponent(scene)}`;
    const signal = AbortSignal.any([controller.signal, this.lifeCtl.signal]);
    try {
      const res = await this.fetchImpl(url, {
        headers: { accept: 'text/event-stream' },
        signal,
      });
      if (!res.ok || !res.body) throw new Error(`local stream ${scene} HTTP ${res.status}`);
      log(`dashrelay local SSE subscribed: scene=${scene}`);
      const parser = new SseDataParser();
      const reader = res.body.getReader();
      const decoder = new TextDecoder();
      // eslint-disable-next-line no-constant-condition
      while (true) {
        const { value, done } = await reader.read();
        if (done) break;
        if (signal.aborted || this.stopped) break;
        const chunk = decoder.decode(value, { stream: true });
        for (const envelope of parser.push(chunk)) {
          if (!this.dashConnected) return; // dashrelay gone; stop pumping
          const sc = (envelope as { scene?: unknown } | null)?.scene;
          if (typeof sc !== 'string') continue;
          this.publishDash({ type: 'frame', envelope });
        }
      }
      if (!signal.aborted) this.retryDashScene(scene);
    } catch (e) {
      if (signal.aborted || this.stopped) return;
      logErr(`dashrelay local SSE scene=${scene} error: ${errMsg(e)}; retrying`);
      this.retryDashScene(scene);
    }
  }

  private retryDashScene(scene: string): void {
    if (this.stopped || !this.dashConnected) return;
    if (this.dashPumpRetryTimers.has(scene)) return;
    const t = setTimeout(() => {
      this.dashPumpRetryTimers.delete(scene);
      if (this.stopped || !this.dashConnected) return;
      this.subscribeDashScene(scene);
    }, this.cfg.localRetryMs);
    this.dashPumpRetryTimers.set(scene, t);
    if (typeof t.unref === 'function') t.unref();
  }

  // ---- DOWN: control (POST to local, reply control-result UP) ----

  private async applyControl(ws: WsLike, id: string | undefined, args: unknown): Promise<void> {
    const op = (args as { op?: unknown } | undefined)?.op;
    log(`control op=${typeof op === 'string' ? op : '?'} id=${id ?? '?'}`);
    // Fail-soft: the local control endpoint never throws (aeon-monitor.rs:4869);
    // any network/parse failure becomes {ok:false,error} so the app degrades.
    let result: { ok?: boolean; status?: unknown; error?: unknown };
    try {
      const res = await this.fetchImpl(`${this.cfg.local}/monitor/control`, {
        method: 'POST',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify(args ?? {}),
        signal: this.lifeCtl.signal,
      });
      result = (await res.json().catch(() => ({ ok: false, error: 'invalid control response' }))) as {
        ok?: boolean;
        status?: unknown;
        error?: unknown;
      };
    } catch (e) {
      result = { ok: false, error: `local control unreachable: ${errMsg(e)}` };
    }
    this.sendUp(ws, {
      type: 'control-result',
      id,
      ok: result.ok,
      status: result.status,
      error: result.error,
    });
    this.controlsApplied++;
    log(`control-result id=${id ?? '?'} ok=${result.ok === true}`);
  }

  // ---- helpers ----

  /** GET local JSON; return null (not throw) when the local monitor is down. */
  private async localGetJson(path: string): Promise<unknown | null> {
    try {
      const res = await this.fetchImpl(`${this.cfg.local}${path}`, {
        headers: { accept: 'application/json' },
        signal: this.lifeCtl.signal,
      });
      if (!res.ok) return null;
      return await res.json();
    } catch {
      return null;
    }
  }

  private sendUp(ws: WsLike, msg: Record<string, unknown>): void {
    if (ws.readyState !== WS_OPEN) return;
    try {
      ws.send(JSON.stringify(msg));
    } catch (e) {
      logErr(`relay send failed (${String(msg.type)}): ${errMsg(e)}`);
    }
  }

  /** Stop polling + drop all local subscriptions (on close / stop). */
  private teardownRelaySideEffects(): void {
    this.stopStatusPolling();
    this.lastStatusJson = '';
    for (const scene of [...this.localSubs.keys()]) this.unsubscribeLocalScene(scene);
  }

  /** Test/inspection accessors. */
  get stats(): {
    framesUp: number;
    controlsApplied: number;
    connected: boolean;
    dashFramesUp: number;
    dashConnected: boolean;
    dashJoined: boolean;
    admitted: boolean;
  } {
    return {
      framesUp: this.framesUp,
      controlsApplied: this.controlsApplied,
      connected: this.connected,
      dashFramesUp: this.dashFramesUp,
      dashConnected: this.dashConnected,
      dashJoined: this.dashJoined,
      admitted: this.admitted,
    };
  }
}

function errMsg(e: unknown): string {
  if (e instanceof Error) return e.message;
  if (e && typeof e === 'object' && 'message' in e) return String((e as { message: unknown }).message);
  return String(e);
}

// --------------------------------------------------------------------------
// PREFLIGHT PARIS GATE prover runner.
//
// Spawns the VERIFIED paris-probe-rknot binary (REAL inference, ~3s), parses the
// verdict from stdout, and returns a structured PreflightResult. Never throws --
// a spawn/timeout failure becomes {pass:false, detail:"probe error: ..."}. The
// runner is injectable (ProbeRunner) so unit tests can supply a fake binary /
// stub WITHOUT mocking the verdict in the shipped path.
// --------------------------------------------------------------------------

/** The raw result of running the prover process (exit code + captured stdout). */
export interface ProbeRun {
  /** Process exit code (0 = the probe's own PASS gate). null when killed. */
  code: number | null;
  /** Combined stdout (the PASS line + top-K table live here). */
  stdout: string;
  /** Combined stderr (probe diagnostics). */
  stderr: string;
}

/**
 * Runs the prover process. Default = execFile the configured probe binary with
 * the FIXED Paris arguments. Injectable so tests can swap in a fake binary
 * runner. MUST resolve (never reject) -- carry any spawn error in `stderr` with a
 * non-zero/null `code`.
 */
export type ProbeRunner = (probeBin: string, model: string, timeoutMs: number) => Promise<ProbeRun>;

/** The FIXED prover argv (Paris gate). Exposed so the help + tests agree. */
export function probeArgs(model: string): string[] {
  return [
    '--dense',
    model,
    '--dense-only',
    '--tokens',
    PREFLIGHT_TOKENS.join(','),
    '--expected',
    String(PREFLIGHT_EXPECTED_TOKEN),
  ];
}

/** Default ProbeRunner: execFile the real paris-probe-rknot binary. */
export const defaultProbeRunner: ProbeRunner = (probeBin, model, timeoutMs) =>
  new Promise<ProbeRun>((resolve) => {
    execFile(
      probeBin,
      probeArgs(model),
      { timeout: timeoutMs, maxBuffer: 16 * 1024 * 1024, killSignal: 'SIGKILL' },
      (err, stdout, stderr) => {
        // execFile gives err.code (process exit code) on a non-zero exit, or an
        // Error with no numeric code on spawn failure (ENOENT) / timeout.
        let code: number | null = 0;
        if (err) {
          const c = (err as { code?: unknown }).code;
          if (typeof c === 'number') code = c;
          else code = null; // spawn failure / killed (ENOENT, timeout)
          const stderrPlus =
            typeof c === 'number'
              ? String(stderr ?? '')
              : `${String(stderr ?? '')}\n${errMsg(err)}`;
          resolve({ code, stdout: String(stdout ?? ''), stderr: stderrPlus });
          return;
        }
        resolve({ code, stdout: String(stdout ?? ''), stderr: String(stderr ?? '') });
      },
    );
  });

/**
 * Parse the prover's stdout into the rank-0 token + logit and the PASS verdict.
 *   PASS line: `PASS: argmax token_id=12095 matches expected 12095`
 *   rank-0:    `  rank 0: token_id= 12095  logit=+16.822826`
 * PASS requires exit code 0 AND a PASS line. The argmax token + logit come from
 * the rank-0 row (falling back to the PASS line's token if the table is absent).
 */
export function parseProbeOutput(run: ProbeRun): { pass: boolean; token: number; logit: number; detail?: string } {
  const out = run.stdout;
  const passLine = /^PASS: argmax token_id=(\d+) matches expected (\d+)/m.exec(out);
  const rank0 = /rank 0:\s*token_id=\s*(\d+)\s+logit=\s*([+-]?[\d.eE]+)/.exec(out);

  const token = rank0
    ? Number.parseInt(rank0[1]!, 10)
    : passLine
      ? Number.parseInt(passLine[1]!, 10)
      : NaN;
  const logit = rank0 ? Number.parseFloat(rank0[2]!) : NaN;

  // PASS = the probe's own gate (exit 0) AND the PASS line is present.
  const pass = run.code === 0 && passLine != null;

  if (pass) return { pass: true, token, logit };

  // Build a useful detail for the FAIL/error path.
  let detail: string;
  if (run.code === null) {
    detail = `probe error: ${firstNonEmptyLine(run.stderr) || 'process killed (timeout or spawn failure)'}`;
  } else if (run.code !== 0) {
    const failLine = /^FAIL:.*/m.exec(out);
    detail = failLine
      ? failLine[0]!.trim()
      : `probe exited ${run.code}: ${firstNonEmptyLine(run.stderr) || firstNonEmptyLine(out) || 'no output'}`;
  } else {
    detail = 'probe exited 0 but no PASS line found';
  }
  return {
    pass: false,
    token: Number.isFinite(token) ? token : 0,
    logit: Number.isFinite(logit) ? logit : 0,
    detail,
  };
}

function firstNonEmptyLine(s: string): string {
  for (const line of s.split('\n')) {
    const t = line.trim();
    if (t) return t;
  }
  return '';
}

const MONITOR_FLOW_SCHEMA = 'gnosis.monitor.v1';
const MONITOR_FLOW_KINDS = ['raster', 'vector', 'embed', 'skymesh'] as const;

/**
 * Coerce a local monitor SSE envelope into a MonitorFlowFrame the codec can
 * carry. The local monitor already emits `gnosis.monitor.v1` envelopes; this
 * fills the numeric width/height/ts the MonitorFlow assert requires (so the
 * frame round-trips through encode/decode) and rejects anything that is not a
 * valid monitor envelope (returns null -> the frame is skipped, never thrown).
 */
export function monitorFlowFrameFromEnvelope(envelope: unknown): MonitorFlowFrame | null {
  if (!envelope || typeof envelope !== 'object') return null;
  const e = envelope as Record<string, unknown>;
  if (e.schemaVersion !== MONITOR_FLOW_SCHEMA) return null;
  if (typeof e.scene !== 'string') return null;
  if (!MONITOR_FLOW_KINDS.includes(e.kind as (typeof MONITOR_FLOW_KINDS)[number])) return null;
  const width = typeof e.width === 'number' ? e.width : 0;
  const height = typeof e.height === 'number' ? e.height : 0;
  const ts = typeof e.ts === 'number' ? e.ts : Date.now();
  // raster frames must carry a string raster payload + numeric dims.
  if (e.kind === 'raster' && (typeof e.raster !== 'string' || width <= 0 || height <= 0)) {
    return null;
  }
  return { ...(e as object), width, height, ts } as MonitorFlowFrame;
}

// --------------------------------------------------------------------------
// MESH QUERY runner.
//
// Ports bin/skymesh-query's VERIFIED path to a factored, unit-testable function:
//   1) tokens given -> use them; else POST /tokenize {text:prompt} (the node's
//      in-tree smoke table may 400 on arbitrary prose -> surface the error).
//   2) feed each prompt token through POST /decode-next (X-Token/X-Position);
//      the reply after the LAST prompt token is the first prediction.
//   3) continue autoregressively to maxTokens or a qwen stop token.
//   4) POST /detokenize {token_ids} (best-effort).
// The fetch is injectable so tests drive a mocked fat-station WITHOUT mocking the
// reply (the decode is REAL against whatever fetch returns). NEVER throws: on no
// node / tokenize 400 / decode error it returns a query-result with `detail`.
// --------------------------------------------------------------------------

/** Args the bridge derives from a DOWN {type:"query",...} message. */
export interface MeshQueryInput {
  /** Prompt text to tokenize via the node (ignored when `tokens` is given). */
  prompt?: string;
  /** Raw prompt token ids (bypass the smoke-tokenizer encode limit). */
  tokens?: number[];
  /** Max new tokens to decode (default QUERY_DEFAULT_MAX_TOKENS). */
  maxTokens?: number;
}

/** The decoded outcome (sans the wire `type`/`id`, which the caller stamps). */
export interface MeshQueryOutcome {
  reply: string;
  promptTokens: number[];
  outTokens: number[];
  elapsedMs: number;
  model: string;
  detail?: string;
  /** True when the reply was served from the GLOBAL ANSWER CACHE (no inference). */
  cached?: boolean;
  /** The mycelial tier that served a cache hit (fano7 / aeon66 / monster196884). */
  cacheTier?: string;
}

/**
 * A served cache hit (the reply replayed without inference). `tier` is the
 * serving tier when the cache surfaces it.
 */
export interface MeshQueryCacheHit {
  outTokens: number[];
  reply: string;
  tier?: string;
}

/**
 * The cache seam runMeshQuery calls when a cache is configured. `lookup` runs
 * AFTER prompt tokens are resolved and BEFORE inference; a non-null return is a
 * VERIFIED hit served verbatim (no decode). `store` runs AFTER a successful
 * decode (a cache MISS path); it is best-effort and must never throw. Both are
 * graceful: an unreachable cache returns null (lookup) / no-ops (store).
 */
export interface MeshQueryCacheHooks {
  lookup: (promptTokens: number[], model: string) => Promise<MeshQueryCacheHit | null>;
  store: (
    promptTokens: number[],
    outTokens: number[],
    reply: string,
    model: string,
  ) => Promise<void>;
}

/** Minimal fetch surface the query runner needs (so tests inject a mock). */
export type QueryFetch = (
  url: string,
  init?: { method?: string; headers?: Record<string, string>; body?: string },
) => Promise<{ ok: boolean; status: number; text: () => Promise<string> }>;

/**
 * Run the REAL decode-next mesh query against a fat-station at `nodeUrl`. The
 * `fetchImpl` defaults to the global fetch; tests inject a mock fat-station. The
 * `modelOverride` (from --model) wins; absent, the node's /status `model` is
 * used; absent that, '' . NEVER throws.
 */
export async function runMeshQuery(
  nodeUrl: string,
  input: MeshQueryInput,
  fetchImpl: QueryFetch,
  modelOverride?: string,
  cacheHooks?: MeshQueryCacheHooks,
): Promise<MeshQueryOutcome> {
  const base = nodeUrl.replace(/\/+$/, '');
  const maxTokens =
    typeof input.maxTokens === 'number' && Number.isFinite(input.maxTokens) && input.maxTokens > 0
      ? Math.floor(input.maxTokens)
      : QUERY_DEFAULT_MAX_TOKENS;
  const stop = new Set<number>(QUERY_STOP_TOKENS);
  const t0 = Date.now();

  // POST helper local to the runner; resolves the parsed JSON or throws a typed
  // error carrying the HTTP status + a short body excerpt (callers catch it).
  const post = async (
    path: string,
    opts: { headers?: Record<string, string>; body?: string } = {},
  ): Promise<unknown> => {
    const res = await fetchImpl(`${base}${path}`, { method: 'POST', ...opts });
    const text = await res.text();
    if (!res.ok) throw new Error(`${path} HTTP ${res.status}: ${text.slice(0, 160)}`);
    try {
      return JSON.parse(text);
    } catch {
      return text;
    }
  };

  // Resolve the model name: --model override wins; else best-effort /status.
  let model = modelOverride ?? '';
  if (!model) {
    try {
      const res = await fetchImpl(`${base}/status`, { method: 'GET' });
      if (res.ok) {
        const s = JSON.parse(await res.text()) as { model?: unknown };
        if (typeof s.model === 'string') model = s.model;
      }
    } catch {
      /* /status is best-effort; the reply still ships with model '' */
    }
  }

  // 1) prompt tokens: raw `tokens` bypass the smoke encoder; else /tokenize.
  let promptTokens: number[];
  if (Array.isArray(input.tokens) && input.tokens.length > 0) {
    promptTokens = input.tokens.filter((n) => Number.isFinite(n));
  } else if (typeof input.prompt === 'string' && input.prompt.length > 0) {
    try {
      const tok = (await post('/tokenize', {
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify({ text: input.prompt }),
      })) as { tokens?: number[]; token_ids?: number[] };
      promptTokens = tok.tokens ?? tok.token_ids ?? [];
    } catch (e) {
      // The node's in-tree tokenizer is tiny; arbitrary prose 400s. Surface it.
      return {
        reply: '',
        promptTokens: [],
        outTokens: [],
        elapsedMs: Date.now() - t0,
        model,
        detail:
          `tokenize failed: ${errMsg(e)} -- the node's in-tree tokenizer is a tiny ` +
          `smoke table; pass raw token ids (tokens) or run the production tokenizer`,
      };
    }
  } else {
    return {
      reply: '',
      promptTokens: [],
      outTokens: [],
      elapsedMs: Date.now() - t0,
      model,
      detail: 'no prompt or tokens given',
    };
  }

  if (promptTokens.length === 0) {
    return {
      reply: '',
      promptTokens: [],
      outTokens: [],
      elapsedMs: Date.now() - t0,
      model,
      detail: 'empty prompt token sequence',
    };
  }

  // GLOBAL ANSWER CACHE lookup (WS-3): with a cache configured, probe it BEFORE
  // inference. A VERIFIED hit is replayed verbatim (no decode at all). The hook
  // is graceful -- an unreachable/erroring cache returns null and we fall through
  // to inference.
  if (cacheHooks) {
    const hit = await cacheHooks.lookup(promptTokens, model);
    if (hit) {
      return {
        reply: hit.reply,
        promptTokens,
        outTokens: hit.outTokens,
        elapsedMs: Date.now() - t0,
        model,
        cached: true,
        ...(hit.tier ? { cacheTier: hit.tier } : {}),
      };
    }
  }

  // 2-3) feed the prompt one token at a time, then continue autoregressively.
  const outTokens: number[] = [];
  try {
    let pos = 0;
    let next: number | null = null;
    for (const t of promptTokens) {
      const r = (await post('/decode-next', {
        headers: { 'X-Token': String(t), 'X-Position': String(pos++) },
      })) as { token_id?: number };
      next = typeof r.token_id === 'number' ? r.token_id : null;
    }
    for (let i = 0; i < maxTokens && next != null && !stop.has(next); i++) {
      outTokens.push(next);
      const r = (await post('/decode-next', {
        headers: { 'X-Token': String(next), 'X-Position': String(pos++) },
      })) as { token_id?: number };
      next = typeof r.token_id === 'number' ? r.token_id : null;
    }
  } catch (e) {
    return {
      reply: '',
      promptTokens,
      outTokens,
      elapsedMs: Date.now() - t0,
      model,
      detail: `decode failed: ${errMsg(e)} -- no inference node at ${base}?`,
    };
  }

  // 4) detokenize (best-effort; the smoke table may not cover all tokens).
  let reply = '';
  try {
    const d = (await post('/detokenize', {
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify({ token_ids: outTokens }),
    })) as { text?: unknown };
    if (typeof d.text === 'string') reply = d.text;
  } catch {
    /* detokenize best-effort; outTokens still carry the answer */
  }

  // GLOBAL ANSWER CACHE store (WS-3): a cache MISS that produced a real answer is
  // offered to the cache. The hook gates on admission + signs + POSTs, and is
  // best-effort (never throws, never blocks the reply on a cache failure). We only
  // store when there is a real answer to replay (non-empty outTokens).
  if (cacheHooks && outTokens.length > 0) {
    await cacheHooks.store(promptTokens, outTokens, reply, model);
  }

  return {
    reply,
    promptTokens,
    outTokens,
    elapsedMs: Date.now() - t0,
    model,
    cached: false,
  };
}

// --------------------------------------------------------------------------
// Entry point (skipped when imported, e.g. by the test).
// --------------------------------------------------------------------------

function main(): void {
  const cfg = configFromArgv(process.argv, process.env);
  if (process.argv.includes('--help') || process.argv.includes('-h')) {
    console.log(
      [
        'skymesh-bridge -- connect the LOCAL aeon-monitor up to the hosted relay',
        '',
        'Usage: bin/skymesh-bridge [--relay <wss-url>] [--local <http-url>]',
        '                          [--token <secret>] [--scenes a,b,c]',
        '                          [--mesh-id <id>] [--node-id <id>]',
        '                          [--dashrelay <wss-base>] [--no-relay]',
        '                          [--model <path>] [--probe-bin <path>]',
        '                          [--node-url <http-url>]',
        '                          [--cache-url <base>] [--no-cache]',
        '',
        `  --relay      relay WSS endpoint   (default ${DEFAULT_RELAY})`,
        `  --local      local monitor base   (default ${DEFAULT_LOCAL})`,
        '  --token      bridge secret        (default env SKYMESH_BRIDGE_TOKEN)',
        `  --scenes     scenes to mirror     (default ${DEFAULT_SCENES})`,
        `  --mesh-id    GLOBAL PARTYLINE mesh (default ${SKYMESH_GLOBAL_MESH_ID})`,
        '  --node-id    this node id         (default os.hostname())',
        '  --dashrelay  OPTIONAL DashRelay wss base (e.g. wss://dashrelay.com)',
        '  --no-relay   skip the hosted relay; publish to DashRelay only',
        `  --model      PARIS-gate model     (default ${DEFAULT_PREFLIGHT_MODEL})`,
        '  --model-id   GLOBAL cache model identity (default model basename, no ext)',
        '  --probe-bin  paris-probe-rknot prover (default the repo release build)',
        `  --node-url   local fat-station for mesh queries (default ${DEFAULT_NODE_URL})`,
        '  --cache-url  GLOBAL ANSWER CACHE base (e.g. https://edgework.ai); empty = OFF',
        '  --no-cache   force the global answer cache OFF (overrides --cache-url)',
        '',
        'GLOBAL PARTYLINE: the relay routes per --mesh-id; every bridge on the',
        'default mesh joins one global room. The upgrade carries ?mesh=&node= and',
        'the first WS message is {type:"hello",nodeId}; returned frames carry',
        'metadata.nodeId. --dashrelay adds an ADDITIVE second uplink to a DashRelay',
        'aeon-flow room (/.aeon/flow, opaque-relay; each frame is a MonitorFlow',
        'Flow frame, binary FlowCodec).',
        '',
        'PREFLIGHT PARIS GATE: on a DOWN {type:"preflight",id} the bridge spawns the',
        'VERIFIED paris-probe-rknot prover (REAL inference, ~3s) against --model,',
        'asserts the next token after "The capital of France is" is " Paris"',
        '(token 12095), and replies {type:"preflight-result",...} UP. A PASS marks',
        'the node admitted (carried in hello/status) -- prove, then join global.',
        '',
        'MESH QUERY: on a DOWN {type:"query",id,prompt?,tokens?,maxTokens} the bridge',
        'runs the REAL per-token autoregressive decode against the local fat-station',
        '(--node-url) over the VERIFIED /tokenize + /decode-next path (NOT /generate),',
        'stopping at a qwen stop token or maxTokens, then /detokenize, and replies',
        '{type:"query-result",id,reply,promptTokens,outTokens,elapsedMs,model,detail?}',
        'UP. The node tokenizer is a tiny smoke table, so arbitrary prose may 400 --',
        'pass raw token ids (tokens) and the detail explains any tokenizer error.',
        '',
        'GLOBAL ANSWER CACHE (--cache-url): when set, a mesh query first looks up the',
        'global answer cache (GET /api/v1/cache/lookup, keyed by the fp48 canonical',
        'hash of the prompt tokens + model + skymesh-query/v1) and serves a VERIFIED',
        'hit (pass + admitted + signed) verbatim with cached:true -- NO inference. On',
        'a miss it runs inference and, ONLY IF this node is admitted, signs + stores',
        'the answer (POST /api/v1/cache/put) and returns cached:false. The cache is',
        'graceful: any cache error falls through to inference and never breaks a query.',
        '',
        'Forwards frame/status/scenes UP and applies control + preflight + query DOWN',
        'over the FIXED uplink protocol; interops with the MonitorRelayDO relay.',
      ].join('\n'),
    );
    return;
  }
  // The hosted-relay path needs the bridge token; the DashRelay-only path does
  // not (the room is the routing key). Require a token only when the relay is on.
  if (!cfg.noRelay && !cfg.token) {
    logErr('no token: set SKYMESH_BRIDGE_TOKEN or pass --token (must match the Worker secret)');
    logErr('(or run DashRelay-only with --no-relay --dashrelay <wss-base>)');
    process.exitCode = 1;
    return;
  }
  if (cfg.noRelay && !cfg.dashrelay) {
    logErr('--no-relay requires --dashrelay <wss-base> (nothing to publish to otherwise)');
    process.exitCode = 1;
    return;
  }
  const bridge = new SkymeshBridge(cfg);
  bridge.start();
  const shutdown = () => {
    bridge.stop();
    process.exit(0);
  };
  process.once('SIGINT', shutdown);
  process.once('SIGTERM', shutdown);
}

// Only run when executed directly (not when imported by the test).
const isMain = (() => {
  try {
    const invoked = process.argv[1] ?? '';
    return invoked.endsWith('skymesh-bridge.mts') || invoked.endsWith('skymesh-bridge.js');
  } catch {
    return false;
  }
})();

if (isMain) {
  try {
    main();
  } catch (e) {
    logErr(errMsg(e));
    process.exitCode = 1;
  }
}
