// @a0n/.ignored_gnosis/rtlsdr-mock-sim/skymesh-serve.mts
#!/usr/bin/env node
/**
 * Skymesh live frame server.
 *
 * A zero-extra-dependency Node HTTP server that streams gnosis.skymesh.v1 frames
 * to the Skymesh web app (built separately, in apps/skymesh -- NOT
 * touched here). The app polls GET /skymesh/frame.json for the latest frame and
 * subscribes to GET /skymesh/stream (Server-Sent Events) for ~4 fps live updates;
 * the UI itself animates at 12 fps from the pushed beat/spectrum state, so the
 * server only needs to push beat transitions + evolving spectra.
 *
 * Three modes:
 *   --mode demo    (default) synthesize the 6-beat sequence as live frames with
 *                  evolving 64-band spectra (moving noise floor + a carrier bin
 *                  that locks). No hardware. Drives the UI with motion.
 *   --mode file    poll a frame.json file (default the canonical skymesh-run.mts
 *                  output) and push it on change -- a real run drives the UI.
 *   --mode capture OPTIONAL/best-effort: shell rtl_sdr -d 0 / -d 1 to grab short
 *                  windows, extractWitness on each, admit, build + push the frame.
 *                  This is the REAL two-NESDR path; if rtl_sdr or the devices are
 *                  unavailable it logs and falls back to demo. It never crashes
 *                  the server.
 *
 * CORS is open (Access-Control-Allow-Origin: *) so the Vite dev server can consume
 * it cross-origin.
 */
import { createServer, type IncomingMessage, type ServerResponse } from 'node:http';
import { createSocket, type Socket as DgramSocket } from 'node:dgram';
import { readFileSync, statSync } from 'node:fs';
import { extname, resolve } from 'node:path';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import {
  admitStation,
  buildStation,
  buildSkymeshFrame,
  AIR_GAP_PROOF,
  SKYMESH_SCHEMA_VERSION,
  DEFAULT_SKYMESH_FRAME_PATH,
  type SkymeshFrame,
  type SkymeshStation,
} from './skymesh.ts';
import {
  extractWitness,
  DEFAULT_WITNESS_OPTS,
} from './skymesh-witness.ts';
import {
  probeRtlSdr,
  captureRtlSdrWindow,
} from './rf-cache-key.ts';
import { PROTOCOL69_BROADCAST_SYMBOL } from '../src/protocol69.ts';
// The canonical broadcast wire is aeon's Flow protocol. Each UDP datagram is one
// FlowCodec-encoded ConsciousTick frame (10-byte header + JSON payload) on the
// Skymesh stream (0x5353). See open-source/aeon/src/flow/FlowCodec.ts (the wire
// authority) and open-source/aeon/src/flow/SkymeshFlow.ts (the frame builder).
import {
  FlowCodec,
  encodeSkymeshFlowFrame,
  SKYMESH_FLOW_STREAM_ID,
  FORK,
  FOLD,
  type SkymeshFlowFrame,
} from '@a0n/aeon/flow';

// --------------------------------------------------------------------------
// CLI
// --------------------------------------------------------------------------

function argValue(name: string, def: string): string {
  const i = process.argv.indexOf(name);
  if (i === -1 || i + 1 >= process.argv.length) return def;
  return process.argv[i + 1]!;
}
function argFlag(name: string): boolean {
  return process.argv.includes(name);
}
/**
 * Resolve a flag that may carry an OPTIONAL value. `--udp` enables UDP and may
 * be followed by a port (`--udp 8725`); a bare `--udp` (or `--udp` followed by
 * another flag) uses the default. Returns the parsed value or `def` when the
 * flag is present without a usable value, or `null` when the flag is absent.
 */
function argOptionalValue(name: string, def: string): string | null {
  const i = process.argv.indexOf(name);
  if (i === -1) return null;
  const next = process.argv[i + 1];
  if (next === undefined || next.startsWith('--')) return def;
  return next;
}
/** Collect every occurrence of a repeatable `name <value>` flag, in order. */
function argValues(name: string): string[] {
  const out: string[] = [];
  for (let i = 0; i < process.argv.length; i++) {
    if (process.argv[i] === name && i + 1 < process.argv.length) {
      out.push(process.argv[i + 1]!);
    }
  }
  return out;
}

type ServeMode = 'demo' | 'file' | 'capture';

const SPECTRUM_BANDS = DEFAULT_WITNESS_OPTS.bands; // 64

// --------------------------------------------------------------------------
// Demo frame generator (pure; the test exercises this directly).
//
// The web-app cycle is six beats. The schema's station states are LISTENING ->
// WITNESS_LOCKED -> TELEPORTED -> DENIED; we add a "WITNESS_LOCKING" beat label
// for the lock-in animation (it still renders as a locked-but-not-admitted
// station, i.e. state WITNESS_LOCKED with witnessLocked true). The last two beats
// rotate the failure cases: DENIED (a tampered/forged key -- same sky, denied)
// and MISS (a different sky -- witnesses disagree). Both yield match:false.
// --------------------------------------------------------------------------

export type DemoBeatLabel =
  | 'LISTENING'
  | 'WITNESS_LOCKING'
  | 'WITNESS_LOCKED'
  | 'TELEPORTED'
  | 'DENIED'
  | 'MISS';

/** The canonical six-beat demo cycle, in order. */
export const DEMO_BEAT_SEQUENCE: readonly DemoBeatLabel[] = [
  'LISTENING',
  'WITNESS_LOCKING',
  'WITNESS_LOCKED',
  'TELEPORTED',
  'DENIED',
  'MISS',
] as const;

/**
 * Synthesize a plausible 64-band spectrum: a slowly-moving noise floor plus a
 * carrier bin that sharpens as the beat progresses. `lockProgress` in [0,1]
 * raises the carrier above the floor; `phase` slides the floor so the UI sees
 * motion every push. `carrierBand` selects which bin carries (two stations of the
 * same sky share it; a MISS station carries a different bin).
 */
export function synthSpectrum(args: {
  readonly phase: number;
  readonly lockProgress: number;
  readonly carrierBand: number;
  readonly bands?: number;
}): number[] {
  const bands = args.bands ?? SPECTRUM_BANDS;
  const carrier = ((args.carrierBand % bands) + bands) % bands;
  const raw = new Array<number>(bands);
  let total = 0;
  for (let b = 0; b < bands; b++) {
    // Moving noise floor: a low, smooth, phase-shifted ripple. Deterministic so
    // the same (phase, beat) always yields the same spectrum (testable).
    const floor =
      0.4 +
      0.15 * Math.sin((b / bands) * Math.PI * 6 + args.phase) +
      0.1 * Math.sin((b / bands) * Math.PI * 13 - args.phase * 1.7);
    // Carrier: a peak at the carrier band, growing with lockProgress, with a
    // little spectral skirt on the neighbours.
    const d = Math.min(Math.abs(b - carrier), bands - Math.abs(b - carrier));
    const carrierGain =
      args.lockProgress > 0 ? 8 * args.lockProgress * Math.exp(-(d * d) / 2) : 0;
    const v = Math.max(0, floor) + carrierGain;
    raw[b] = v;
    total += v;
  }
  // Normalize to a shape vector (matches the witness spectrum convention).
  const out = new Array<number>(bands);
  for (let b = 0; b < bands; b++) out[b] = total > 0 ? raw[b]! / total : 0;
  return out;
}

/** Build one station record for the demo, by-passing real witness extraction. */
function demoStation(args: {
  readonly id: string;
  readonly witness: number;
  readonly witnessLocked: boolean;
  readonly referenceWitness: number;
  readonly spectrum: number[];
  /** Forge the projection key so admission fails closed (DENIED beat). */
  readonly broadcastSymbol?: number;
  /**
   * Force the displayed state to WITNESS_LOCKED: the station has locked a witness
   * but is not yet attempting admission against a reference. We run admission with
   * witnessLocked=false (so admitStation returns LISTENING, never DENIED) and then
   * present witnessLocked=true to buildStation, which promotes LISTENING ->
   * WITNESS_LOCKED. This is the pre-teleport beat (locked, no token).
   */
  readonly displayLockedOnly?: boolean;
}): SkymeshStation {
  const lockedForAdmission = args.displayLockedOnly ? false : args.witnessLocked;
  const admission = admitStation({
    witness: args.witness,
    witnessLocked: lockedForAdmission,
    referenceWitness: args.referenceWitness,
    broadcastSymbol: args.broadcastSymbol,
  });
  return buildStation({
    id: args.id,
    witness: args.witness,
    witnessLocked: args.witnessLocked,
    spectrum: args.spectrum,
    admission,
  });
}

// Shared-sky witness for the two stations (top-3 dominant bands encoded 8 bits
// each, ascending -- see skymesh-witness.encodeBands). The carrier band is one
// of them so the synthesized spectrum and the witness tell the same story.
const SHARED_CARRIER_BAND = 30;
const SHARED_WITNESS = (10 << 16) | (20 << 8) | SHARED_CARRIER_BAND;
// A MISS station hears a different sky: an entirely different dominant-band set.
const MISS_CARRIER_BAND = 47;
const MISS_WITNESS = (3 << 16) | (15 << 8) | MISS_CARRIER_BAND;

/**
 * Produce the gnosis.skymesh.v1 frame for one beat of the demo cycle. `tick`
 * advances the moving noise floor so successive pushes animate. Pure + total:
 * the test asserts the schema shape, spectrum length, and the beat progression
 * (TELEPORTED -> match true; DENIED/MISS -> match false).
 */
export function demoFrameForBeat(beat: DemoBeatLabel, tick: number): SkymeshFrame {
  const phase = (tick % 360) * 0.12;

  switch (beat) {
    case 'LISTENING': {
      // No lock yet: both stations scanning, no token. Low, moving floor.
      const specA = synthSpectrum({ phase, lockProgress: 0, carrierBand: SHARED_CARRIER_BAND });
      const specB = synthSpectrum({ phase: phase + 0.6, lockProgress: 0, carrierBand: SHARED_CARRIER_BAND });
      const a = demoStation({ id: 'station-a', witness: SHARED_WITNESS, witnessLocked: false, referenceWitness: SHARED_WITNESS, spectrum: specA });
      const b = demoStation({ id: 'station-b', witness: SHARED_WITNESS, witnessLocked: false, referenceWitness: SHARED_WITNESS, spectrum: specB });
      return buildSkymeshFrame([a, b]);
    }
    case 'WITNESS_LOCKING': {
      // Carrier rising; locked but not yet admitting against the sky (no ref yet).
      const lp = 0.45;
      const specA = synthSpectrum({ phase, lockProgress: lp, carrierBand: SHARED_CARRIER_BAND });
      const specB = synthSpectrum({ phase: phase + 0.6, lockProgress: lp, carrierBand: SHARED_CARRIER_BAND });
      // Locked off the air but not yet admitting against a reference -> the
      // displayed state is WITNESS_LOCKED (no token yet).
      const a = demoStation({ id: 'station-a', witness: SHARED_WITNESS, witnessLocked: true, referenceWitness: SHARED_WITNESS, spectrum: specA, displayLockedOnly: true });
      const b = demoStation({ id: 'station-b', witness: SHARED_WITNESS, witnessLocked: true, referenceWitness: SHARED_WITNESS, spectrum: specB, displayLockedOnly: true });
      return buildSkymeshFrame([a, b]);
    }
    case 'WITNESS_LOCKED': {
      // Both fully locked on the same carrier; the shared answer has not
      // teleported yet (WITNESS_LOCKED, no token).
      const specA = synthSpectrum({ phase, lockProgress: 0.85, carrierBand: SHARED_CARRIER_BAND });
      const specB = synthSpectrum({ phase: phase + 0.6, lockProgress: 0.85, carrierBand: SHARED_CARRIER_BAND });
      const a = demoStation({ id: 'station-a', witness: SHARED_WITNESS, witnessLocked: true, referenceWitness: SHARED_WITNESS, spectrum: specA, displayLockedOnly: true });
      const b = demoStation({ id: 'station-b', witness: SHARED_WITNESS, witnessLocked: true, referenceWitness: SHARED_WITNESS, spectrum: specB, displayLockedOnly: true });
      return buildSkymeshFrame([a, b]);
    }
    case 'TELEPORTED': {
      // Both admit against each other's witness: 66 XOR 7 = 69, cached answer
      // served in lockstep. match true.
      const specA = synthSpectrum({ phase, lockProgress: 1, carrierBand: SHARED_CARRIER_BAND });
      const specB = synthSpectrum({ phase: phase + 0.6, lockProgress: 1, carrierBand: SHARED_CARRIER_BAND });
      const a = demoStation({ id: 'station-a', witness: SHARED_WITNESS, witnessLocked: true, referenceWitness: SHARED_WITNESS, spectrum: specA });
      const b = demoStation({ id: 'station-b', witness: SHARED_WITNESS, witnessLocked: true, referenceWitness: SHARED_WITNESS, spectrum: specB });
      return buildSkymeshFrame([a, b]);
    }
    case 'DENIED': {
      // Station B heard the SAME sky (witness agrees) but presents a forged
      // protocol69 key (broadcast symbol 67): projection 67 XOR 7 = 64, not 69 ->
      // fail closed. A teleports, B DENIED. match false.
      const specA = synthSpectrum({ phase, lockProgress: 1, carrierBand: SHARED_CARRIER_BAND });
      const specB = synthSpectrum({ phase: phase + 0.6, lockProgress: 1, carrierBand: SHARED_CARRIER_BAND });
      const a = demoStation({ id: 'station-a', witness: SHARED_WITNESS, witnessLocked: true, referenceWitness: SHARED_WITNESS, spectrum: specA });
      const b = demoStation({ id: 'station-b', witness: SHARED_WITNESS, witnessLocked: true, referenceWitness: SHARED_WITNESS, spectrum: specB, broadcastSymbol: 67 });
      return buildSkymeshFrame([a, b]);
    }
    case 'MISS': {
      // Station B hears a DIFFERENT sky -> different witness -> no agreement ->
      // neither is admitted to the shared answer. match false.
      const specA = synthSpectrum({ phase, lockProgress: 1, carrierBand: SHARED_CARRIER_BAND });
      const specB = synthSpectrum({ phase: phase + 0.6, lockProgress: 1, carrierBand: MISS_CARRIER_BAND });
      const a = demoStation({ id: 'station-a', witness: SHARED_WITNESS, witnessLocked: true, referenceWitness: MISS_WITNESS, spectrum: specA });
      const b = demoStation({ id: 'station-b', witness: MISS_WITNESS, witnessLocked: true, referenceWitness: SHARED_WITNESS, spectrum: specB });
      return buildSkymeshFrame([a, b]);
    }
  }
}

/**
 * Stateful demo frame generator. Each call advances the cycle: it holds each
 * beat for `holdTicks` pushes (so a beat visibly animates before transitioning),
 * then steps to the next beat, wrapping after MISS back to LISTENING.
 */
export function makeDemoGenerator(holdTicks = 8): () => SkymeshFrame {
  let tick = 0;
  return () => {
    const beatIndex = Math.floor(tick / Math.max(1, holdTicks)) % DEMO_BEAT_SEQUENCE.length;
    const beat = DEMO_BEAT_SEQUENCE[beatIndex]!;
    const frame = demoFrameForBeat(beat, tick);
    tick++;
    return frame;
  };
}

/**
 * The full six-beat cycle as a frame array (one frame per beat, no animation
 * sub-ticks). Useful for the test, which asserts the progression end to end.
 */
export function demoCycleFrames(tick = 0): SkymeshFrame[] {
  return DEMO_BEAT_SEQUENCE.map((beat, i) => demoFrameForBeat(beat, tick + i));
}

// --------------------------------------------------------------------------
// File mode: poll a frame.json and push on change.
// --------------------------------------------------------------------------

function readFrameFile(path: string): SkymeshFrame | null {
  try {
    const raw = readFileSync(path, 'utf8');
    const parsed = JSON.parse(raw) as SkymeshFrame;
    if (parsed?.schemaVersion !== SKYMESH_SCHEMA_VERSION) return null;
    return parsed;
  } catch {
    return null;
  }
}

// --------------------------------------------------------------------------
// Capture mode: best-effort two-NESDR path. Never crashes the server.
// --------------------------------------------------------------------------

interface CaptureConfig {
  readonly binary: string;
  readonly centerFrequencyHz: number;
  readonly sampleRateHz: number;
  readonly samplePairs: number;
  readonly lockConfidence: number;
}

/**
 * Attempt one capture-mode frame off two real NESDRs (-d 0 / -d 1). Returns null
 * if any device/binary is unavailable or a capture fails; the caller falls back.
 */
function captureFrameOnce(cfg: CaptureConfig): SkymeshFrame | null {
  const probe = probeRtlSdr();
  if (!probe.healthy || probe.devices.length < 2) {
    return null;
  }
  const witnesses: { id: string; witness: number; locked: boolean; spectrum: number[] }[] = [];
  for (const dev of ['0', '1']) {
    const out = join(tmpdir(), `skymesh-capture-d${dev}.u8`);
    const cap = captureRtlSdrWindow({
      binary: cfg.binary,
      outputPath: out,
      centerFrequencyHz: cfg.centerFrequencyHz,
      sampleRateHz: cfg.sampleRateHz,
      samplePairs: cfg.samplePairs,
      deviceIndex: dev,
    });
    if (!cap.ok) return null;
    let iq: Uint8Array;
    try {
      iq = new Uint8Array(readFileSync(out));
    } catch {
      return null;
    }
    const w = extractWitness(iq, { sampleRateHz: cfg.sampleRateHz });
    witnesses.push({
      id: dev === '0' ? 'station-a' : 'station-b',
      witness: w.witness,
      locked: w.confidence >= cfg.lockConfidence,
      spectrum: w.spectrum,
    });
  }
  const [a, b] = witnesses;
  if (!a || !b) return null;
  const admA = admitStation({
    witness: a.witness,
    witnessLocked: a.locked,
    referenceWitness: b.witness,
    broadcastSymbol: PROTOCOL69_BROADCAST_SYMBOL,
  });
  const admB = admitStation({
    witness: b.witness,
    witnessLocked: b.locked,
    referenceWitness: a.witness,
    broadcastSymbol: PROTOCOL69_BROADCAST_SYMBOL,
  });
  const stationA = buildStation({ id: a.id, witness: a.witness, witnessLocked: a.locked, spectrum: a.spectrum, admission: admA });
  const stationB = buildStation({ id: b.id, witness: b.witness, witnessLocked: b.locked, spectrum: b.spectrum, admission: admB });
  return buildSkymeshFrame([stationA, stationB]);
}

// --------------------------------------------------------------------------
// UDP datagram publisher over the aeon Flow wire (node:dgram, no new deps).
//
// Each new frame is emitted as exactly ONE FlowCodec-encoded ConsciousTick
// datagram (10-byte big-endian header + JSON payload) on the Skymesh Flow
// stream (0x5353) -- the canonical aeon Flow protocol, NOT an ad-hoc raw-JSON
// scheme. The JSON payload is the SAME gnosis.skymesh.v1 frame the SSE clients
// receive. One Flow frame per datagram (no MTU fragmentation for this hose).
//
// Multicast (the "same sky" broadcast: every listener on the group hears the
// identical datagram) is the default fan-out; unicast targets may be added or
// used instead. The publisher NEVER takes down the HTTP/SSE server: bind/send
// failures are logged and swallowed.
//
// Wire authority: open-source/aeon/src/flow/FlowCodec.ts (10-byte header --
// u16 streamId BE, u32 sequence BE, u8 flags, u24 length BE) and SkymeshFlow.ts
// (SKYMESH_FLOW_STREAM_ID = 0x5353, schema gnosis.skymesh.v1).
// --------------------------------------------------------------------------

/** UDP wire defaults (FIXED contract; other agents depend on these). */
export const UDP_DEFAULT_PORT = 8725;
export const UDP_DEFAULT_GROUP = '239.69.0.66';
export const UDP_DEFAULT_TTL = 1;
/**
 * Max UDP payload over IPv4 (65535 total - 20 IP header - 8 UDP header). A
 * Flow frame whose encoded bytes exceed this is skipped with a warning --
 * never truncated. Measured on the full FlowCodec datagram (header + payload).
 */
export const UDP_MAX_DATAGRAM_BYTES = 65507;

/**
 * Flow stream id for Skymesh broadcasts (re-exported for subscribers/tests
 * that want the contract value without importing @a0n/aeon directly).
 * @see open-source/aeon/src/flow/SkymeshFlow.ts
 */
export const UDP_FLOW_STREAM_ID = SKYMESH_FLOW_STREAM_ID; // 0x5353

/**
 * Flow flags for a Skymesh broadcast frame: FORK | FOLD (= 0x05). Each
 * datagram is a self-contained fork-then-fold tick (the air-gapped stations
 * fork off the same sky and fold to the same answer). FIXED contract.
 */
export const UDP_FLOW_FLAGS = FORK | FOLD; // 0x05

interface UdpTarget {
  readonly host: string;
  readonly port: number;
}

export interface UdpConfig {
  readonly enabled: boolean;
  readonly port: number;
  readonly group: string;
  readonly ttl: number;
  /** Unicast targets (in addition to the multicast group). */
  readonly targets: readonly UdpTarget[];
}

/** Parse a repeatable `host:port` target. Throws on a malformed entry. */
export function parseUdpTarget(raw: string): UdpTarget {
  const idx = raw.lastIndexOf(':');
  if (idx <= 0 || idx === raw.length - 1) {
    throw new Error(`--udp-target must be host:port (got: ${raw})`);
  }
  const host = raw.slice(0, idx);
  const port = Number.parseInt(raw.slice(idx + 1), 10);
  if (!Number.isInteger(port) || port < 1 || port > 65535) {
    throw new Error(`--udp-target port out of range in: ${raw}`);
  }
  return { host, port };
}

/**
 * A small frame publisher over a single UDP socket. Create once at startup via
 * `createUdpPublisher`; call `.publish(frame)` on every frame push; `.close()`
 * on shutdown. Returns `null` from the factory if UDP is disabled or the socket
 * cannot be created -- callers must treat UDP as best-effort and keep serving.
 */
export interface UdpPublisher {
  publish(frame: SkymeshFrame): void;
  close(): void;
}

export function createUdpPublisher(cfg: UdpConfig): UdpPublisher | null {
  if (!cfg.enabled) return null;

  // The Flow encoder is pure + synchronous (JS path; WASM is a perf-only
  // optimization we do not need here). One codec, reused for every frame.
  const codec = FlowCodec.createSync();
  // Monotonic Flow sequence counter (0, 1, 2, ...). u32 wraps at 2^32 to stay
  // within the Flow header's u32 sequence field.
  let sequence = 0;

  let socket: DgramSocket;
  try {
    socket = createSocket({ type: 'udp4', reuseAddr: true });
  } catch (e) {
    console.error(
      `[skymesh-serve] udp: failed to create dgram socket: ` +
        `${e instanceof Error ? e.message : String(e)} (udp disabled, http/sse unaffected)`,
    );
    return null;
  }

  // Never let an asynchronous socket error crash the process; log and keep the
  // HTTP/SSE server running. A udp error is not fatal to the server.
  socket.on('error', (err) => {
    console.error(`[skymesh-serve] udp: socket error: ${err.message} (continuing)`);
  });

  let multicastReady = false;
  // Bind so we can set the multicast TTL; bind is async, so set TTL on listening.
  try {
    socket.bind(() => {
      try {
        socket.setMulticastTTL(cfg.ttl);
        multicastReady = true;
      } catch (e) {
        console.error(
          `[skymesh-serve] udp: setMulticastTTL failed: ` +
            `${e instanceof Error ? e.message : String(e)} ` +
            `(multicast group sends may be skipped; unicast targets still used)`,
        );
      }
    });
  } catch (e) {
    console.error(
      `[skymesh-serve] udp: bind failed: ${e instanceof Error ? e.message : String(e)} ` +
        `(udp disabled, http/sse unaffected)`,
    );
    try {
      socket.close();
    } catch {
      /* ignore */
    }
    return null;
  }

  function send(payload: Buffer, host: string, port: number): void {
    socket.send(payload, port, host, (err) => {
      if (err) {
        console.error(
          `[skymesh-serve] udp: send to ${host}:${port} failed: ${err.message} (continuing)`,
        );
      }
    });
  }

  return {
    publish(frame: SkymeshFrame): void {
      // Encode ONE aeon Flow ConsciousTick frame: streamId 0x5353, a monotonic
      // sequence, flags FORK|FOLD (0x05), payload = utf8(JSON of the same frame
      // the SSE clients receive). encodeSkymeshFlowFrame owns the schema check
      // and the payload bytes; we pin the FIXED sequence + flags contract.
      const flowFrame = encodeSkymeshFlowFrame(
        frame as unknown as SkymeshFlowFrame,
        { sequence: sequence >>> 0, flags: UDP_FLOW_FLAGS },
      );
      sequence = (sequence + 1) >>> 0; // advance + wrap at u32, monotonic
      const encoded = codec.encode(flowFrame); // 10-byte header + payload
      const datagram = Buffer.from(
        encoded.buffer,
        encoded.byteOffset,
        encoded.byteLength,
      );
      // Honest size guard on the FULL Flow datagram: never silently truncate.
      // Skymesh frames are a few KB so this should never fire, but skip + warn
      // if it ever does. One Flow frame per datagram (no fragmentation).
      if (datagram.byteLength > UDP_MAX_DATAGRAM_BYTES) {
        console.error(
          `[skymesh-serve] udp: Flow frame is ${datagram.byteLength} bytes ` +
            `(> ${UDP_MAX_DATAGRAM_BYTES} UDP/IPv4 limit); skipping this datagram ` +
            `(never truncated). HTTP/SSE unaffected.`,
        );
        return;
      }
      // Multicast fan-out (the skymesh broadcast).
      if (multicastReady) {
        send(datagram, cfg.group, cfg.port);
      }
      // Unicast targets.
      for (const t of cfg.targets) {
        send(datagram, t.host, t.port);
      }
    },
    close(): void {
      try {
        socket.close();
      } catch {
        /* already closed */
      }
    },
  };
}

// --------------------------------------------------------------------------
// HTTP / SSE server.
// --------------------------------------------------------------------------

const CORS_HEADERS = {
  'Access-Control-Allow-Origin': '*',
  'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
  'Access-Control-Allow-Headers': 'Content-Type',
} as const;

interface ServerOptions {
  readonly mode: ServeMode;
  readonly port: number;
  readonly host: string;
  readonly intervalMs: number;
  readonly framePath: string;
  readonly capture: CaptureConfig;
  readonly udp: UdpConfig;
}

interface ClientEntropySample {
  readonly receivedAt: number;
  readonly remote: string;
  readonly payload: unknown;
}

interface ExhaustEntry {
  readonly id: string;
  readonly type: 'system' | 'navigation' | 'tool_invoke' | 'tool_result' | 'biometric';
  readonly timestamp: number;
  readonly attribution: 'on-record' | 'background' | 'deep-background' | 'off-record';
  readonly content: unknown;
  readonly visible: boolean;
  readonly metadata?: Record<string, unknown>;
}

interface SkymeshQueryResult {
  readonly reply: string;
  readonly promptTokens: number[];
  readonly outTokens: number[];
  readonly elapsedMs: number;
  readonly model: string;
  readonly detail?: string;
  readonly cached?: boolean;
  readonly cacheTier?: string;
  readonly usage?: unknown;
  readonly agentic?: unknown;
}

interface ThothGenerationResult {
  readonly promptTokens: number[];
  readonly outTokens: number[];
  readonly reply: string;
  readonly detail: string;
  readonly usage?: unknown;
  readonly agentic?: unknown;
}

interface OpenAiChatMessage {
  readonly role?: unknown;
  readonly content?: unknown;
}

interface OpenAiChatRequest {
  readonly model?: unknown;
  readonly messages?: unknown;
  readonly max_tokens?: unknown;
  readonly stream?: unknown;
}

interface OpenAiChatCompletionResponse {
  readonly model?: unknown;
  readonly choices?: unknown;
  readonly usage?: unknown;
  readonly agentic?: unknown;
}

function readRequestBody(req: IncomingMessage, limit = 64 * 1024): Promise<string> {
  return new Promise((resolveBody, reject) => {
    let body = '';
    req.setEncoding('utf8');
    req.on('data', (chunk: string) => {
      body += chunk;
      if (body.length > limit) {
        reject(new Error('request body too large'));
        req.destroy();
      }
    });
    req.on('end', () => resolveBody(body));
    req.on('error', reject);
  });
}

const QUERY_MODELS = [
  {
    id: 'thoth',
    label: 'Thoth',
    detail: 'Agentic OpenAI-compatible Thoth route through distributed-inference-host.',
  },
  {
    id: 'paris-gate',
    label: 'PARIS gate',
    detail: 'Text completion test: "The capital of France is" -> " Paris" / token 12095.',
  },
  {
    id: 'open=a',
    label: 'open=a SSE',
    detail: 'AP-local query path over the same server that emits Skymesh SSE frames.',
  },
  {
    id: 'browser-entropy',
    label: 'Browser entropy',
    detail: 'Summarizes Safari client entropy received at POST /client/entropy.',
  },
  {
    id: 'skymesh-demo',
    label: 'Skymesh demo',
    detail: 'Deterministic local proof response from the current gnosis.skymesh.v1 frame.',
  },
] as const;

function tokenShape(text: string, maxTokens: number): number[] {
  const out: number[] = [];
  const limit = Math.max(1, Math.min(maxTokens, 64));
  for (let i = 0; i < text.length && out.length < limit; i++) {
    out.push((text.charCodeAt(i) + i * 17) & 0xffff);
  }
  return out;
}

function buildQueryResult(args: {
  readonly model: string;
  readonly prompt: string;
  readonly maxTokens: number;
  readonly latest: SkymeshFrame;
  readonly entropySamples: readonly ClientEntropySample[];
  readonly startedAt: number;
}): SkymeshQueryResult {
  const latestEntropy = args.entropySamples[args.entropySamples.length - 1];
  const entropyPayload = latestEntropy?.payload as
    | { session?: unknown; samples?: unknown; hardwareConcurrency?: unknown; screen?: unknown }
    | undefined;
  const sampleCount = Array.isArray(entropyPayload?.samples) ? entropyPayload.samples.length : 0;
  const entropyLine = latestEntropy
    ? `client entropy: ${args.entropySamples.length} post(s), latest remote ${latestEntropy.remote}, ${sampleCount} sample(s)`
    : 'client entropy: no Safari entropy post received yet';
  const stationLine = args.latest.stations
    .map((s) => `${s.id}:${s.state}:w${s.witness}`)
    .join(' ');
  const model = QUERY_MODELS.some((m) => m.id === args.model) ? args.model : 'open=a';
  const reply =
    model === 'thoth'
      ? `Thoth received: ${args.prompt || '(empty prompt)'}. Attach a live Thoth/fat-station upstream to replace this AP-local fallback with fresh inference.`
      : model === 'paris-gate'
      ? ' Paris'
      : model === 'browser-entropy'
      ? `${entropyLine}. WebGPU/CPU/motion availability is harvested in-browser when Safari exposes it.`
      : model === 'skymesh-demo'
        ? `frame ${args.latest.schemaVersion}: match=${args.latest.match}; ${stationLine}`
        : `open=a on skymesh AP: SSE frames live at /skymesh/stream; ${entropyLine}; current ${stationLine}`;
  const clipped =
    model === 'paris-gate'
      ? reply
      : reply.split(/\s+/).slice(0, Math.max(1, args.maxTokens)).join(' ');
  return {
    reply: clipped,
    promptTokens:
      args.prompt.trim() === 'The capital of France is'
        ? [785, 6722, 315, 9625, 374]
        : tokenShape(args.prompt, Math.max(1, args.prompt.length)),
    outTokens: model === 'paris-gate' ? [12095] : tokenShape(clipped, args.maxTokens),
    elapsedMs: Date.now() - args.startedAt,
    model,
    detail:
      model === 'paris-gate'
        ? 'PARIS quality gate PASS: expected next token " Paris" / 12095; node admitted by AP fallback'
        : model === 'thoth'
          ? 'Thoth AP route selected; no live Thoth upstream configured on this server yet'
        : 'served by skymesh-serve AP fallback; no monitor bind required',
    cached: false,
  };
}

async function postJson(url: string, body: unknown, timeoutMs: number): Promise<unknown> {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  try {
    const res = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
      signal: controller.signal,
    });
    const text = await res.text();
    if (!res.ok) {
      throw new Error(`${url} returned ${res.status}: ${text}`);
    }
    return text ? JSON.parse(text) : {};
  } finally {
    clearTimeout(timer);
  }
}

async function decodeNext(base: string, token: number, position: number): Promise<number> {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), 180_000);
  try {
    const res = await fetch(`${base}/decode-next`, {
      method: 'POST',
      headers: {
        'X-Token': String(token),
        'X-Position': String(position),
      },
      body: '',
      signal: controller.signal,
    });
    const text = await res.text();
    if (!res.ok) {
      throw new Error(`/decode-next returned ${res.status}: ${text}`);
    }
    const json = text ? JSON.parse(text) as { token_id?: unknown } : {};
    if (typeof json.token_id !== 'number') {
      throw new Error('/decode-next response missing token_id');
    }
    return json.token_id;
  } finally {
    clearTimeout(timer);
  }
}

async function tokenizeText(base: string, text: string): Promise<number[]> {
  const tok = await postJson(`${base}/tokenize`, { text }, 30_000) as { tokens?: unknown };
  if (!Array.isArray(tok.tokens)) throw new Error('/tokenize response missing tokens');
  const tokens = tok.tokens.filter((n): n is number => typeof n === 'number');
  if (tokens.length === 0) throw new Error('/tokenize returned no tokens');
  return tokens;
}

async function qwenSentenceCloseTokenIds(base: string): Promise<number[]> {
  const closers = await Promise.allSettled(['.', '!', '?', '.\n', '!\n', '?\n'].map((text) => tokenizeText(base, text)));
  const ids = new Set<number>();
  for (const result of closers) {
    if (result.status !== 'fulfilled') continue;
    for (const token of result.value) ids.add(token);
  }
  return [...ids];
}

function firstCompleteSentence(text: string): string {
  const trimmed = text.trimStart();
  const match = trimmed.match(/^(.+?[.!?])(?=\s|$)/s);
  return match?.[1] ?? text;
}

function qwenChatPrompt(userPrompt: string): string {
  return [
    '<|im_start|>system',
    'Answer briefly in one complete sentence.',
    '<|im_end|>',
    '<|im_start|>user',
    userPrompt,
    '<|im_end|>',
    '<|im_start|>assistant',
    '',
  ].join('\n');
}

function thothPromptText(prompt: string): { text: string; mode: 'raw-paris-gate' | 'qwen-chat' } {
  if (prompt.trim() === 'The capital of France is') {
    return { text: prompt, mode: 'raw-paris-gate' };
  }
  return { text: qwenChatPrompt(prompt), mode: 'qwen-chat' };
}

function isParisSanityPrompt(prompt: string): boolean {
  const normalized = prompt.trim().toLowerCase();
  return normalized === 'the capital of france is' || normalized === 'what is paris?';
}

function passesParisSanity(prompt: string, reply: string): boolean {
  if (!isParisSanityPrompt(prompt)) return true;
  const normalized = reply.toLowerCase();
  return normalized.includes('paris') || normalized.includes('france');
}

async function generateViaThothFatStation(prompt: string, maxTokens: number): Promise<ThothGenerationResult> {
  const base = process.env.SKYMESH_THOTH_URL || process.env.FAT_STATION_URL || 'http://127.0.0.1:8786';
  const prepared = thothPromptText(prompt);
  const promptTokens = await tokenizeText(base, prepared.text);
  const sentenceCloseTokenIds = await qwenSentenceCloseTokenIds(base);
  const stopTokenIds = new Set([2, 151_643, 151_645]);
  const sentenceCloseIds = new Set(sentenceCloseTokenIds);
  const outTokens: number[] = [];
  let feedToken = promptTokens[0]!;
  let pos = 0;
  const totalSteps = promptTokens.length + Math.max(1, maxTokens) - 1;
  for (let step = 0; step < totalSteps; step++) {
    const next = await decodeNext(base, feedToken, pos);
    if (pos + 1 >= promptTokens.length) {
      if (stopTokenIds.has(next)) break;
      outTokens.push(next);
      if (sentenceCloseIds.has(next)) break;
      if (outTokens.length >= maxTokens) break;
    }
    pos++;
    feedToken = pos < promptTokens.length ? promptTokens[pos]! : next;
  }

  let reply = '';
  let source = 'detokenize-unavailable';
  try {
    const detok = await postJson(`${base}/detokenize`, { token_ids: outTokens }, 30_000) as {
      text?: unknown;
      source?: unknown;
    };
    reply = typeof detok.text === 'string' ? detok.text : '';
    source = typeof detok.source === 'string' ? detok.source : 'detokenize';
  } catch (err) {
    reply = outTokens.map((t) => String(t)).join(' ');
    source = err instanceof Error ? err.message : String(err);
  }

  return {
    promptTokens,
    outTokens,
    reply: prepared.mode === 'qwen-chat' ? firstCompleteSentence(reply) : reply,
    detail: `Thoth via distributed-inference fat-station ${base}; generation source /decode-next with ${prepared.mode} prompt and sentence-boundary stop; detokenize source ${source}`,
  };
}

async function generateViaThothOpenAiHost(prompt: string, maxTokens: number): Promise<ThothGenerationResult> {
  const base = (
    process.env.SKYMESH_THOTH_OPENAI_URL ||
    process.env.GNOSIS_BACKEND_URL ||
    'http://127.0.0.1:8787'
  ).replace(/\/+$/, '');
  try {
    const body = await postJson(
      `${base}/v1/chat/completions`,
      {
        model: 'thoth',
        messages: [{ role: 'user', content: prompt }],
        max_tokens: Math.max(1, maxTokens),
        temperature: 0,
        auto_tools: true,
        execute_tools: true,
      },
      180_000,
    ) as OpenAiChatCompletionResponse;
    const choices = Array.isArray(body.choices) ? body.choices : [];
    const first = choices[0] as { message?: { content?: unknown } } | undefined;
    const reply = typeof first?.message?.content === 'string' ? first.message.content : '';
    if (!reply) throw new Error('distributed-inference-host returned no assistant content');
    if (
      process.env.SKYMESH_ALLOW_DIRECT_FAT_STATION_FALLBACK === '1' &&
      !passesParisSanity(prompt, reply)
    ) {
      const fallback = await generateViaThothFatStation(prompt, maxTokens);
      return {
        ...fallback,
        detail:
          `${fallback.detail}; fallback used because distributed-inference-host failed Paris sanity ` +
          `with reply: ${JSON.stringify(reply.slice(0, 160))}`,
        usage: body.usage,
        agentic: body.agentic,
      };
    }
    return {
      promptTokens: [],
      outTokens: [],
      reply,
      detail: `Thoth via distributed-inference-host OpenAI server ${base}; agentic=true; tools requested`,
      usage: body.usage,
      agentic: body.agentic,
    };
  } catch (err) {
    if (process.env.SKYMESH_ALLOW_DIRECT_FAT_STATION_FALLBACK === '1') {
      const fallback = await generateViaThothFatStation(prompt, maxTokens);
      return {
        ...fallback,
        detail:
          `${fallback.detail}; fallback used because distributed-inference-host failed: ` +
          `${err instanceof Error ? err.message : String(err)}`,
      };
    }
    throw err;
  }
}

function messagesToPrompt(messages: readonly OpenAiChatMessage[]): string {
  if (messages.length === 1) {
    const only = messages[0]!;
    if (
      (only.role === undefined || only.role === 'user') &&
      typeof only.content === 'string'
    ) {
      return only.content;
    }
  }
  return messages
    .map((message) => {
      const role = typeof message.role === 'string' ? message.role : 'user';
      const content =
        typeof message.content === 'string'
          ? message.content
          : Array.isArray(message.content)
            ? message.content
                .map((part) => {
                  if (typeof part === 'string') return part;
                  if (part && typeof part === 'object' && typeof (part as { text?: unknown }).text === 'string') {
                    return (part as { text: string }).text;
                  }
                  return '';
                })
                .join('')
            : '';
      return `${role}: ${content}`;
    })
    .join('\n') + '\nassistant:';
}

function openAiCompletionBody(args: {
  readonly model: string;
  readonly promptTokens: readonly number[];
  readonly outTokens: readonly number[];
  readonly content: string;
  readonly elapsedMs: number;
}): string {
  return JSON.stringify({
    id: `chatcmpl-${Date.now()}`,
    object: 'chat.completion',
    created: Math.floor(Date.now() / 1000),
    model: args.model,
    choices: [
      {
        index: 0,
        message: { role: 'assistant', content: args.content },
        finish_reason: 'stop',
      },
    ],
    usage: {
      prompt_tokens: args.promptTokens.length,
      completion_tokens: args.outTokens.length,
      total_tokens: args.promptTokens.length + args.outTokens.length,
    },
    _gnosis: {
      agentic: false,
      elapsed_ms: args.elapsedMs,
      upstream: process.env.SKYMESH_THOTH_URL || process.env.FAT_STATION_URL || 'http://127.0.0.1:8786',
    },
  });
}

function sendOpenAiChunk(res: ServerResponse, payload: unknown): void {
  res.write(`data: ${payload === '[DONE]' ? '[DONE]' : JSON.stringify(payload)}\n\n`);
}

function skymeshAppUrl(req: IncomingMessage): string {
  const hostHeader = req.headers.host ?? '127.0.0.1:8724';
  return `http://${hostHeader}/app/?source=live&live=${encodeURIComponent('/skymesh')}`;
}

function portalHtml(appUrl: string): string {
  return `<!doctype html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta http-equiv="refresh" content="0; url=${appUrl}">
  <title>Skymesh</title>
  <style>
    body{margin:0;min-height:100vh;display:grid;place-items:center;background:#0b0d10;color:#cfd6dd;font-family:-apple-system,BlinkMacSystemFont,"Segoe UI",sans-serif}
    main{width:min(30rem,calc(100vw - 2rem));border:1px solid #353c46;border-left:3px solid #ffaf00;background:#14181d;padding:1rem}
    h1{margin:0 0 .5rem;font-size:1.35rem}
    p{margin:.4rem 0;color:#8a929c;line-height:1.45}
    a{display:inline-block;margin-top:.75rem;color:#050607;background:#5fffff;border:1px solid #5fffff;border-radius:.5rem;padding:.65rem .9rem;text-decoration:none;font-weight:700}
    code{color:#ffaf00}
  </style>
</head>
<body>
  <main>
    <h1>Skymesh AP</h1>
    <p>Opening the phone-facing Skymesh app with live SSE frames and AP-local query enabled.</p>
    <p><code>open=a</code> · <code>/skymesh/stream</code> · <code>/skymesh/query</code></p>
    <a href="${appUrl}">Open Skymesh</a>
  </main>
  <script>location.replace(${JSON.stringify(appUrl)});</script>
</body>
</html>`;
}

const APP_DIST_DIR = resolve(process.cwd(), '../../apps/skymesh/dist/client');

function contentTypeFor(path: string): string {
  switch (extname(path)) {
    case '.html':
      return 'text/html; charset=utf-8';
    case '.css':
      return 'text/css; charset=utf-8';
    case '.js':
      return 'text/javascript; charset=utf-8';
    case '.json':
      return 'application/json';
    case '.svg':
      return 'image/svg+xml';
    case '.png':
      return 'image/png';
    case '.ico':
      return 'image/x-icon';
    default:
      return 'application/octet-stream';
  }
}

function serveBuiltApp(url: string, res: ServerResponse): boolean {
  const relative =
    url === '/app' || url === '/app/'
      ? 'index.html'
      : url.startsWith('/app/')
        ? decodeURIComponent(url.slice('/app/'.length))
        : url.startsWith('/assets/')
          ? decodeURIComponent(url.slice(1))
          : '';
  if (!relative || relative.includes('..')) return false;
  const filePath = resolve(APP_DIST_DIR, relative);
  if (!filePath.startsWith(APP_DIST_DIR)) return false;
  try {
    const body = readFileSync(filePath);
    res.writeHead(200, {
      ...CORS_HEADERS,
      'Content-Type': contentTypeFor(filePath),
      'Cache-Control': relative === 'index.html' ? 'no-store' : 'no-cache',
    });
    res.end(body);
    return true;
  } catch {
    return false;
  }
}

function startServer(opts: ServerOptions): void {
  // The latest frame any endpoint will serve. Seeded immediately so frame.json
  // is never empty before the first push.
  let latest: SkymeshFrame = demoFrameForBeat('LISTENING', 0);
  const sseClients = new Set<ServerResponse>();
  const exhaustClients = new Set<ServerResponse>();
  const entropySamples: ClientEntropySample[] = [];
  const exhaustEntries: ExhaustEntry[] = [];

  // Optional UDP datagram side-channel. Created once at startup; null when the
  // socket could not be created or --udp was not requested.
  const udpPublisher = createUdpPublisher(opts.udp);

  function broadcast(frame: SkymeshFrame): void {
    latest = frame;
    const payload = `data: ${JSON.stringify(frame)}\n\n`;
    for (const res of sseClients) {
      res.write(payload);
    }
    // Send the same frame as one UDP datagram at the same moment SSE is pushed.
    if (udpPublisher) {
      udpPublisher.publish(frame);
    }
  }

  function appendExhaust(entry: Omit<ExhaustEntry, 'id' | 'timestamp' | 'attribution' | 'visible'> & {
    readonly attribution?: ExhaustEntry['attribution'];
    readonly visible?: boolean;
  }): ExhaustEntry {
    const full: ExhaustEntry = {
      id: `${Date.now().toString(36)}-${Math.random().toString(36).slice(2)}`,
      timestamp: Date.now(),
      attribution: entry.attribution ?? 'background',
      visible: entry.visible ?? true,
      type: entry.type,
      content: entry.content,
      metadata: entry.metadata,
    };
    exhaustEntries.push(full);
    while (exhaustEntries.length > 512) exhaustEntries.shift();
    const payload = `data: ${JSON.stringify(full)}\n\n`;
    for (const res of exhaustClients) {
      res.write(payload);
    }
    return full;
  }

  // ---- Frame source per mode ----
  const demoGen = makeDemoGenerator();
  let effectiveMode = opts.mode;
  let fileLastMtimeMs = -1;

  // File mode: load the current frame synchronously at boot so frame.json is the
  // real frame on the very first request (not the seeded LISTENING placeholder).
  if (opts.mode === 'file') {
    try {
      const st = statSync(opts.framePath);
      const frame = readFrameFile(opts.framePath);
      if (frame) {
        fileLastMtimeMs = st.mtimeMs;
        latest = frame;
      }
    } catch {
      // Not present yet; the poll picks it up when skymesh-run.mts writes it.
    }
  }

  function tick(): void {
    if (effectiveMode === 'demo') {
      broadcast(demoGen());
      return;
    }
    if (effectiveMode === 'file') {
      try {
        const st = statSync(opts.framePath);
        if (st.mtimeMs !== fileLastMtimeMs) {
          const frame = readFrameFile(opts.framePath);
          if (frame) {
            fileLastMtimeMs = st.mtimeMs;
            broadcast(frame);
          }
        }
      } catch {
        // File not present yet; keep the last frame, do not crash.
      }
      return;
    }
    if (effectiveMode === 'capture') {
      const frame = captureFrameOnce(opts.capture);
      if (frame) {
        broadcast(frame);
      } else {
        console.error(
          `[skymesh-serve] capture failed this window (device busy / no energy / ` +
            `binary error); falling back to demo for this push.`,
        );
        broadcast(demoGen());
      }
    }
  }

  const server = createServer((req: IncomingMessage, res: ServerResponse) => {
    const url = (req.url ?? '/').split('?')[0];

    if (req.method === 'OPTIONS') {
      res.writeHead(204, CORS_HEADERS);
      res.end();
      return;
    }

    if (url === '/healthz') {
      res.writeHead(200, { ...CORS_HEADERS, 'Content-Type': 'text/plain' });
      res.end('ok');
      return;
    }

    if (url === '/skymesh/exhaust.json') {
      const parsedUrl = new URL(req.url ?? '/', `http://${req.headers.host ?? '127.0.0.1'}`);
      const since = Number(parsedUrl.searchParams.get('since') ?? '0');
      res.writeHead(200, {
        ...CORS_HEADERS,
        'Content-Type': 'application/json',
        'Cache-Control': 'no-store',
      });
      res.end(JSON.stringify({
        ok: true,
        entries: exhaustEntries.filter((entry) => !Number.isFinite(since) || entry.timestamp > since).slice(-80),
      }));
      return;
    }

    if (url === '/skymesh/exhaust') {
      res.writeHead(200, {
        ...CORS_HEADERS,
        'Content-Type': 'text/event-stream',
        'Cache-Control': 'no-cache',
        Connection: 'keep-alive',
      });
      res.write('retry: 2000\n\n');
      for (const entry of exhaustEntries.slice(-24)) {
        res.write(`data: ${JSON.stringify(entry)}\n\n`);
      }
      exhaustClients.add(res);
      req.on('close', () => {
        exhaustClients.delete(res);
      });
      return;
    }

    if (req.method === 'GET' && serveBuiltApp(url, res)) {
      return;
    }

    if (
      req.method === 'GET' &&
      (
        url === '/' ||
        url === '/hotspot-detect.html' ||
        url === '/library/test/success.html' ||
        url === '/generate_204' ||
        url === '/gen_204' ||
        url === '/connecttest.txt' ||
        url === '/ncsi.txt'
      )
    ) {
      const appUrl = skymeshAppUrl(req);
      res.writeHead(200, {
        ...CORS_HEADERS,
        'Content-Type': 'text/html; charset=utf-8',
        'Cache-Control': 'no-store',
      });
      res.end(portalHtml(appUrl));
      return;
    }

    if (url === '/client/entropy' && req.method === 'POST') {
      void readRequestBody(req)
        .then(async (body) => {
          let payload: unknown = null;
          try {
            payload = body ? JSON.parse(body) : null;
          } catch {
            res.writeHead(400, { ...CORS_HEADERS, 'Content-Type': 'application/json' });
            res.end(JSON.stringify({ ok: false, error: 'invalid_json' }));
            return;
          }
          entropySamples.push({
            receivedAt: Date.now(),
            remote: req.socket.remoteAddress ?? 'unknown',
            payload,
          });
          while (entropySamples.length > 512) entropySamples.shift();
          res.writeHead(200, { ...CORS_HEADERS, 'Content-Type': 'application/json' });
          res.end(JSON.stringify({ ok: true, samples: entropySamples.length }));
        })
        .catch((err: unknown) => {
          res.writeHead(400, { ...CORS_HEADERS, 'Content-Type': 'application/json' });
          res.end(JSON.stringify({ ok: false, error: err instanceof Error ? err.message : String(err) }));
        });
      return;
    }

    if (url === '/client/entropy.json') {
      res.writeHead(200, {
        ...CORS_HEADERS,
        'Content-Type': 'application/json',
        'Cache-Control': 'no-store',
      });
      res.end(JSON.stringify({
        count: entropySamples.length,
        latest: entropySamples.slice(-16),
      }));
      return;
    }

    if (url === '/skymesh/models') {
      res.writeHead(200, {
        ...CORS_HEADERS,
        'Content-Type': 'application/json',
        'Cache-Control': 'no-store',
      });
      res.end(JSON.stringify({ ok: true, models: QUERY_MODELS }));
      return;
    }

    if (url === '/v1/models' && req.method === 'GET') {
      res.writeHead(200, {
        ...CORS_HEADERS,
        'Content-Type': 'application/json',
        'Cache-Control': 'no-store',
      });
      res.end(JSON.stringify({
        object: 'list',
        data: QUERY_MODELS.map((model) => ({
          id: model.id,
          object: 'model',
          created: 0,
          owned_by: 'skymesh',
        })),
      }));
      return;
    }

    if (url === '/v1/chat/completions' && req.method === 'POST') {
      const startedAt = Date.now();
      void readRequestBody(req)
        .then(async (body) => {
          let parsed: OpenAiChatRequest;
          try {
            parsed = body ? JSON.parse(body) as OpenAiChatRequest : {};
          } catch {
            res.writeHead(400, { ...CORS_HEADERS, 'Content-Type': 'application/json' });
            res.end(JSON.stringify({ error: { message: 'invalid_json', type: 'invalid_request_error' } }));
            return;
          }
          if (!Array.isArray(parsed.messages)) {
            res.writeHead(400, { ...CORS_HEADERS, 'Content-Type': 'application/json' });
            res.end(JSON.stringify({ error: { message: 'messages must be an array', type: 'invalid_request_error' } }));
            return;
          }

          const model = typeof parsed.model === 'string' ? parsed.model : 'thoth';
          const maxTokens =
            typeof parsed.max_tokens === 'number' && Number.isFinite(parsed.max_tokens)
              ? Math.max(1, Math.min(256, Math.floor(parsed.max_tokens)))
              : 64;
          const prompt = messagesToPrompt(parsed.messages as OpenAiChatMessage[]);
          const invoke = appendExhaust({
            type: 'tool_invoke',
            content: {
              tool: 'openai/chat.completions',
              phase: 'invoke',
              args: { model, max_tokens: maxTokens, stream: parsed.stream === true },
            },
            metadata: { remote: req.socket.remoteAddress ?? 'unknown' },
          });

          try {
            const generated =
              model === 'thoth'
                ? await generateViaThothOpenAiHost(prompt, maxTokens)
                : (() => {
                    const fallback = buildQueryResult({
                      model,
                      prompt,
                      maxTokens,
                      latest,
                      entropySamples,
                      startedAt,
                    });
                    return {
                      promptTokens: fallback.promptTokens,
                      outTokens: fallback.outTokens,
                      reply: fallback.reply,
                      detail: fallback.detail ?? 'skymesh fallback',
                    };
                  })();
            appendExhaust({
              type: 'tool_result',
              content: {
                tool: 'openai/chat.completions',
                phase: 'result',
                result: {
                  model,
                  reply: generated.reply,
                  promptTokens: generated.promptTokens,
                  outTokens: generated.outTokens,
                  elapsedMs: Date.now() - startedAt,
                  detail: generated.detail,
                  usage: generated.usage,
                  agentic: generated.agentic,
                },
                duration: Date.now() - startedAt,
                invokeId: invoke.id,
              },
              metadata: { remote: req.socket.remoteAddress ?? 'unknown' },
            });

            if (parsed.stream === true) {
              res.writeHead(200, {
                ...CORS_HEADERS,
                'Content-Type': 'text/event-stream',
                'Cache-Control': 'no-cache',
                Connection: 'keep-alive',
              });
              const id = `chatcmpl-${Date.now()}`;
              sendOpenAiChunk(res, {
                id,
                object: 'chat.completion.chunk',
                created: Math.floor(Date.now() / 1000),
                model,
                choices: [{ index: 0, delta: { role: 'assistant' }, finish_reason: null }],
              });
              sendOpenAiChunk(res, {
                id,
                object: 'chat.completion.chunk',
                created: Math.floor(Date.now() / 1000),
                model,
                choices: [{ index: 0, delta: { content: generated.reply }, finish_reason: null }],
              });
              sendOpenAiChunk(res, {
                id,
                object: 'chat.completion.chunk',
                created: Math.floor(Date.now() / 1000),
                model,
                choices: [{ index: 0, delta: {}, finish_reason: 'stop' }],
              });
              sendOpenAiChunk(res, '[DONE]');
              res.end();
              return;
            }

            res.writeHead(200, { ...CORS_HEADERS, 'Content-Type': 'application/json' });
            res.end(openAiCompletionBody({
              model,
              promptTokens: generated.promptTokens,
              outTokens: generated.outTokens,
              content: generated.reply,
              elapsedMs: Date.now() - startedAt,
            }));
          } catch (err) {
            appendExhaust({
              type: 'tool_result',
              content: {
                tool: 'openai/chat.completions',
                phase: 'result',
                result: { error: err instanceof Error ? err.message : String(err) },
                duration: Date.now() - startedAt,
                invokeId: invoke.id,
              },
              metadata: { remote: req.socket.remoteAddress ?? 'unknown' },
            });
            res.writeHead(502, { ...CORS_HEADERS, 'Content-Type': 'application/json' });
            res.end(JSON.stringify({
              error: {
                message: err instanceof Error ? err.message : String(err),
                type: 'upstream_error',
              },
            }));
          }
        })
        .catch((err: unknown) => {
          res.writeHead(400, { ...CORS_HEADERS, 'Content-Type': 'application/json' });
          res.end(JSON.stringify({ error: { message: err instanceof Error ? err.message : String(err), type: 'invalid_request_error' } }));
        });
      return;
    }

    if (url === '/skymesh/query' && req.method === 'POST') {
      const startedAt = Date.now();
      void readRequestBody(req)
        .then(async (body) => {
          let parsed: {
            prompt?: unknown;
            tokens?: unknown;
            maxTokens?: unknown;
            model?: unknown;
            clientEntropy?: unknown;
          };
          try {
            parsed = body ? JSON.parse(body) : {};
          } catch {
            res.writeHead(400, { ...CORS_HEADERS, 'Content-Type': 'application/json' });
            res.end(JSON.stringify({ ok: false, error: 'invalid_json' }));
            return;
          }
          const prompt =
            typeof parsed.prompt === 'string'
              ? parsed.prompt
              : Array.isArray(parsed.tokens)
                ? parsed.tokens.join(',')
                : '';
          const maxTokens =
            typeof parsed.maxTokens === 'number' && Number.isFinite(parsed.maxTokens)
              ? Math.max(1, Math.min(256, Math.floor(parsed.maxTokens)))
              : 48;
          const model = typeof parsed.model === 'string' ? parsed.model : 'open=a';
          const invoke = appendExhaust({
            type: 'tool_invoke',
            content: {
              tool: `skymesh/query/${model}`,
              phase: 'invoke',
              args: { prompt, maxTokens },
            },
            metadata: { remote: req.socket.remoteAddress ?? 'unknown' },
          });
          if (parsed.clientEntropy !== undefined) {
            entropySamples.push({
              receivedAt: Date.now(),
              remote: req.socket.remoteAddress ?? 'unknown',
              payload: parsed.clientEntropy,
            });
            while (entropySamples.length > 512) entropySamples.shift();
            appendExhaust({
              type: 'biometric',
              content: parsed.clientEntropy,
              metadata: { remote: req.socket.remoteAddress ?? 'unknown' },
            });
          }
          let result: SkymeshQueryResult;
          if (model === 'thoth') {
            try {
              const generated = await generateViaThothOpenAiHost(prompt, maxTokens);
              result = {
                reply: generated.reply,
                promptTokens: generated.promptTokens,
                outTokens: generated.outTokens,
                elapsedMs: Date.now() - startedAt,
                model: 'thoth',
                detail: generated.detail,
                ...(generated.usage ? { usage: generated.usage } : {}),
                ...(generated.agentic ? { agentic: generated.agentic } : {}),
                cached: false,
              };
            } catch (err) {
              result = {
                reply: '',
                promptTokens: [],
                outTokens: [],
                elapsedMs: Date.now() - startedAt,
                model: 'thoth',
                detail: `Thoth/distributed-inference-host query failed: ${err instanceof Error ? err.message : String(err)}`,
                cached: false,
              };
            }
          } else {
            result = buildQueryResult({
              model,
              prompt,
              maxTokens,
              latest,
              entropySamples,
              startedAt,
            });
          }
          appendExhaust({
            type: 'tool_result',
            content: {
              tool: `skymesh/query/${model}`,
              phase: 'result',
              result,
              duration: result.elapsedMs,
              invokeId: invoke.id,
            },
            metadata: { remote: req.socket.remoteAddress ?? 'unknown' },
          });
          res.writeHead(200, { ...CORS_HEADERS, 'Content-Type': 'application/json' });
          res.end(JSON.stringify({ ok: true, result }));
        })
        .catch((err: unknown) => {
          res.writeHead(400, { ...CORS_HEADERS, 'Content-Type': 'application/json' });
          res.end(JSON.stringify({ ok: false, error: err instanceof Error ? err.message : String(err) }));
        });
      return;
    }

    if (url === '/skymesh/frame.json') {
      res.writeHead(200, {
        ...CORS_HEADERS,
        'Content-Type': 'application/json',
        'Cache-Control': 'no-store',
      });
      res.end(JSON.stringify(latest));
      return;
    }

    if (url === '/skymesh/stream') {
      res.writeHead(200, {
        ...CORS_HEADERS,
        'Content-Type': 'text/event-stream',
        'Cache-Control': 'no-cache',
        Connection: 'keep-alive',
      });
      // Push the current frame immediately so a fresh subscriber renders at once.
      res.write(`retry: ${opts.intervalMs}\n\n`);
      res.write(`data: ${JSON.stringify(latest)}\n\n`);
      sseClients.add(res);
      req.on('close', () => {
        sseClients.delete(res);
      });
      return;
    }

    res.writeHead(404, { ...CORS_HEADERS, 'Content-Type': 'application/json' });
    res.end(JSON.stringify({ error: 'not_found', path: url }));
  });

  server.listen(opts.port, opts.host, () => {
    const base = `http://${opts.host}:${opts.port}`;
    // Probe for NESDRs AFTER the listener is up, so capture mode never delays the
    // server coming online (the probe is a blocking spawnSync). If hardware is
    // missing, fall back to the demo generator.
    if (effectiveMode === 'capture') {
      const probe = probeRtlSdr();
      if (!probe.healthy || probe.devices.length < 2) {
        console.error(
          `[skymesh-serve] capture mode requested but rtl_sdr/devices unavailable ` +
            `(healthy=${probe.healthy}, devices=${probe.devices.length}); ` +
            `falling back to --mode demo. This path needs two real NESDRs.`,
        );
        effectiveMode = 'demo';
      } else {
        console.error(
          `[skymesh-serve] capture mode: ${probe.devices.length} RTL-SDR device(s) detected; ` +
            `capturing from -d 0 and -d 1. NOTE: rtl_sdr capture is synchronous and ` +
            `blocks the push loop for the capture window; this is the real hardware path.`,
        );
      }
    }

    console.log(`[skymesh-serve] mode=${effectiveMode} interval=${opts.intervalMs}ms`);
    console.log(`[skymesh-serve] listening on ${base}`);
    console.log(`[skymesh-serve]   frame:  GET ${base}/skymesh/frame.json`);
    console.log(`[skymesh-serve]   stream: GET ${base}/skymesh/stream  (Server-Sent Events)`);
    console.log(`[skymesh-serve]   query:  POST ${base}/skymesh/query | GET /skymesh/models`);
    console.log(`[skymesh-serve]   health: GET ${base}/healthz`);
    console.log(`[skymesh-serve] entropy: POST ${base}/client/entropy | GET /client/entropy.json`);
    if (effectiveMode === 'file') {
      console.log(`[skymesh-serve]   tailing frame file: ${opts.framePath}`);
    }
    if (udpPublisher) {
      console.log(
        `[skymesh-serve] udp: broadcasting gnosis.skymesh.v1 over aeon Flow ` +
          `(stream 0x${SKYMESH_FLOW_STREAM_ID.toString(16)}) -> ` +
          `${opts.udp.group}:${opts.udp.port} (ttl ${opts.udp.ttl})`,
      );
      if (opts.udp.targets.length > 0) {
        const list = opts.udp.targets.map((t) => `${t.host}:${t.port}`).join(', ');
        console.log(`[skymesh-serve] udp: unicast targets -> ${list}`);
      }
    }

    // Start pushing once the server is live (and after the capture probe).
    const timer = setInterval(tick, opts.intervalMs);
    if (typeof timer.unref === 'function') timer.unref();
  });

  // Graceful shutdown: close the dgram socket so it never leaks on exit.
  const shutdown = (): void => {
    if (udpPublisher) udpPublisher.close();
    try {
      server.close();
    } catch {
      /* ignore */
    }
  };
  process.once('SIGINT', () => {
    shutdown();
    process.exit(0);
  });
  process.once('SIGTERM', () => {
    shutdown();
    process.exit(0);
  });
}

// --------------------------------------------------------------------------
// Entry point (skipped when imported, e.g. by the test).
// --------------------------------------------------------------------------

function parseMode(raw: string): ServeMode {
  if (raw === 'demo' || raw === 'file' || raw === 'capture') return raw;
  throw new Error(`--mode must be demo, file, or capture (got: ${raw})`);
}

function main(): void {
  const mode = parseMode(argValue('--mode', 'demo'));
  const port = Number.parseInt(argValue('--port', '8722'), 10);
  const host = argValue('--host', '127.0.0.1');
  const intervalMs = Number.parseInt(argValue('--interval-ms', '250'), 10);
  const framePath = resolve(argValue('--frame', DEFAULT_SKYMESH_FRAME_PATH));
  const capture: CaptureConfig = {
    binary: argValue('--rtl-sdr-bin', 'rtl_sdr'),
    centerFrequencyHz: Number.parseInt(argValue('--center-hz', '100000000'), 10),
    sampleRateHz: Number.parseInt(argValue('--iq-rate', '2400000'), 10),
    samplePairs: Number.parseInt(argValue('--sample-pairs', '262144'), 10),
    lockConfidence: Number.parseFloat(argValue('--lock-confidence', '0.12')),
  };

  // UDP datagram side-channel (additive; off unless --udp is present).
  const udpPortRaw = argOptionalValue('--udp', String(UDP_DEFAULT_PORT));
  const udp: UdpConfig = {
    enabled: udpPortRaw !== null,
    port:
      udpPortRaw !== null
        ? Number.parseInt(udpPortRaw, 10) || UDP_DEFAULT_PORT
        : UDP_DEFAULT_PORT,
    group: argValue('--udp-group', UDP_DEFAULT_GROUP),
    ttl: Number.parseInt(argValue('--udp-ttl', String(UDP_DEFAULT_TTL)), 10),
    targets: argValues('--udp-target').map(parseUdpTarget),
  };

  if (argFlag('--help') || argFlag('-h')) {
    console.log(
      [
        'skymesh-serve -- stream gnosis.skymesh.v1 frames to the web app',
        '',
        'Usage: tsx skymesh-serve.mts [--mode demo|file|capture] [--port 8722]',
        '',
        '  --mode demo      (default) synthesize the 6-beat live demo (no hardware)',
        '  --mode file      poll a frame.json and push on change (--frame <path>)',
        '  --mode capture   best-effort two-NESDR capture; falls back to demo',
        '  --port <n>       listen port (default 8722)',
        '  --host <addr>    bind address (default 127.0.0.1)',
        '  --interval-ms    push interval (default 250 = ~4 fps)',
        '  --frame <path>   frame.json for --mode file (default ' + DEFAULT_SKYMESH_FRAME_PATH + ')',
        '',
        'UDP datagram side-channel over the aeon Flow wire (additive; one FlowCodec',
        'frame per datagram, stream 0x' + SKYMESH_FLOW_STREAM_ID.toString(16) + ', schema gnosis.skymesh.v1):',
        '  --udp [port]        enable UDP Flow publishing (default port ' + UDP_DEFAULT_PORT + ')',
        '  --udp-group <grp>   multicast group (default ' + UDP_DEFAULT_GROUP + ')',
        '  --udp-ttl <n>       multicast TTL (default ' + UDP_DEFAULT_TTL + ' = link-local)',
        '  --udp-target h:p    unicast target host:port (repeatable; in addition to multicast)',
        '',
        'Endpoints: GET /skymesh/frame.json, GET /skymesh/stream (SSE), GET /healthz',
      ].join('\n'),
    );
    return;
  }

  startServer({ mode, port, host, intervalMs, framePath, capture, udp });
}

// Only run the server when executed directly (tsx/node), not when imported.
const isMain = (() => {
  try {
    const invoked = process.argv[1] ? resolve(process.argv[1]) : '';
    return invoked.endsWith('skymesh-serve.mts') || invoked.endsWith('skymesh-serve.js');
  } catch {
    return false;
  }
})();

if (isMain) {
  try {
    main();
  } catch (e) {
    console.error(e instanceof Error ? e.message : String(e));
    process.exitCode = 1;
  }
}
