// @a0n/.ignored_gnosis/rtlsdr-mock-sim/whip-diversity.ts
/**
 * Whip Wave fork/race/fold diversity (EXPERIMENT_005). Fork: transmit the same
 * FEC-coded frame N times (time diversity). Race: the channel corrupts each
 * replica independently. Fold: the receiver decodes each and accepts the first
 * that passes CRC -- the surviving path. This is the redundancy layer on top of
 * per-replica FEC; deep fades that wipe one replica rarely wipe all N.
 */
import {
  demodulateFrame,
  frameSampleCount,
  modulateFrame,
  type DemodResult,
  type MfskOptions,
} from './mfsk-modem';

const DC = 128;

export interface WhipReplicasResult {
  readonly iq: Uint8Array;
  readonly replicas: number;
  readonly frameSamples: number;
  readonly gapSamples: number;
}

/** Build N back-to-back replicas of the frame separated by silence gaps. */
export function buildWhipReplicasIq(
  payload: Uint8Array,
  o: MfskOptions,
  replicas: number,
  gapSamples = 256,
): WhipReplicasResult {
  if (replicas < 1) throw new Error('buildWhipReplicasIq: replicas >= 1');
  const frame = modulateFrame(payload, o);
  const frameSamples = frame.length / 2;
  const gap = new Uint8Array(gapSamples * 2).fill(DC);
  const out = new Uint8Array((frame.length + gap.length) * replicas);
  let o2 = 0;
  for (let r = 0; r < replicas; r++) {
    out.set(frame, o2);
    o2 += frame.length;
    out.set(gap, o2);
    o2 += gap.length;
  }
  return { iq: out, replicas, frameSamples, gapSamples };
}

export interface WhipFoldResult {
  readonly ok: boolean;
  readonly payload: Uint8Array | null;
  readonly winnerReplica: number;
  readonly attempts: readonly DemodResult[];
}

/**
 * Fold: try each replica window in order; return the first CRC-valid decode.
 * `payloadLen` is the expected payload length so we can locate each replica
 * window deterministically (the cache-key length is known to both ends).
 */
export function foldWhipReplicas(
  iq: Uint8Array,
  o: MfskOptions,
  replicas: number,
  payloadLen: number,
  gapSamples = 256,
): WhipFoldResult {
  const frameSamples = frameSampleCount(payloadLen, o);
  const stride = frameSamples + gapSamples;
  const attempts: DemodResult[] = [];
  for (let r = 0; r < replicas; r++) {
    const from = Math.max(0, r * stride - gapSamples);
    const res = demodulateFrame(iq, o, from, frameSamples);
    attempts.push(res);
    if (res.ok) {
      return { ok: true, payload: res.payload, winnerReplica: r, attempts };
    }
  }
  return { ok: false, payload: null, winnerReplica: -1, attempts };
}
