// @a0n/.ignored_gnosis/rtlsdr-mock-sim/wrapper-mesh-delegation.mjs
/**
 * Monster-mesh CLI delegation: routing, profiles, probes, adapters only.
 * No distributed mesh algorithms — those live in distributed-inference* crates.
 */

import { constants as fsConstants } from 'node:fs';
import { accessSync, existsSync } from 'node:fs';
import { spawnSync } from 'node:child_process';
import { dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import { createRequire } from 'node:module';

/** @param {string} path */
function canExecute(path) {
  try {
    accessSync(path, fsConstants.X_OK);
    return true;
  } catch {
    return false;
  }
}

/**
 * @param {string[]} argv argv after "monster-mesh" (subcommand stripped)
 * @returns {{ rest: string[]; meshBackend: string | null; profile: string | null }}
 */
export function parseMonsterMeshArgv(argv) {
  const rest = [];
  let meshBackend = null;
  let profile = null;
  for (let i = 0; i < argv.length; i += 1) {
    const a = argv[i];
    if (a === '--mesh-backend' && i + 1 < argv.length) {
      meshBackend = argv[i + 1] ?? null;
      i += 1;
      continue;
    }
    if (a === '--profile' && i + 1 < argv.length) {
      profile = argv[i + 1] ?? null;
      i += 1;
      continue;
    }
    rest.push(a);
  }
  return { rest, meshBackend, profile };
}

/**
 * @param {string | null | undefined} profile
 * @returns {Record<string, string>}
 */
export function profileEnvForDelegation(profile) {
  if (!profile) return {};
  switch (profile) {
    case 'ism433':
      return {
        GNOSIS_MESH_PROFILE: 'ism433',
        GNOSIS_MESH_CENTER_MHZ: '433.92',
      };
    case 'ism915':
      return {
        GNOSIS_MESH_PROFILE: 'ism915',
        GNOSIS_MESH_CENTER_MHZ: '915',
      };
    case 'nwr162':
      return {
        GNOSIS_MESH_PROFILE: 'nwr162',
        GNOSIS_MESH_CENTER_MHZ: '162.475',
      };
    default:
      return {};
  }
}

/**
 * @param {string} gnosisRoot open-source/gnosis absolute path
 * @returns {string[]}
 */
export function aeonMonitorBinaryCandidates(gnosisRoot) {
  const fromEnv = process.env.GNOSIS_AEON_MONITOR_BIN?.trim();
  if (fromEnv) {
    return [resolve(fromEnv)];
  }
  return [
    resolve(gnosisRoot, 'distributed-inference/target/release/aeon-monitor'),
    resolve(gnosisRoot, 'target/release/aeon-monitor'),
  ];
}

/**
 * @param {string} gnosisRoot
 * @returns {string | null} first executable path
 */
export function findAeonMonitorBinary(gnosisRoot) {
  for (const p of aeonMonitorBinaryCandidates(gnosisRoot)) {
    if (canExecute(p)) return p;
  }
  return null;
}

/**
 * @param {string | null} explicit from --mesh-backend
 * @param {string} gnosisRoot
 * @returns {'host' | 'native' | 'service'}
 */
export function resolveMonsterMeshBackend(explicit, gnosisRoot) {
  const trimmed = explicit?.trim() ?? '';
  const fromEnv = (process.env.GNOSIS_MONSTER_MESH_BACKEND ?? '').trim().toLowerCase();

  let chosen = trimmed.toLowerCase();
  if (!chosen && fromEnv) chosen = fromEnv;
  if (!chosen || chosen === 'auto') {
    return findAeonMonitorBinary(gnosisRoot) ? 'native' : 'host';
  }
  if (chosen === 'host' || chosen === 'native' || chosen === 'service') {
    return chosen;
  }
  throw new Error(
    `Invalid --mesh-backend "${explicit}" (expected host|native|service|auto, or GNOSIS_MONSTER_MESH_BACKEND)`,
  );
}

/**
 * @returns {{ ok: true } | { ok: false; reason: string }}
 */
export function probeHostBackend() {
  const meshEntry = resolve(
    dirname(fileURLToPath(import.meta.url)),
    '../distributed-inference-host/src/bin/rknot-mesh-smoke.ts',
  );
  if (!existsSync(meshEntry)) {
    return {
      ok: false,
      reason: `host smoke entrypoint missing: ${meshEntry}`,
    };
  }
  if (!process.env.MESH_STATION_URLS?.trim()) {
    return {
      ok: false,
      reason:
        'host backend requires MESH_STATION_URLS (comma-separated fat-station base URLs)',
    };
  }
  return { ok: true };
}

/**
 * @param {string} gnosisRoot
 * @returns {{ ok: true; bin: string } | { ok: false; reason: string }}
 */
export function probeNativeBackend(gnosisRoot) {
  const bin = findAeonMonitorBinary(gnosisRoot);
  if (!bin) {
    return {
      ok: false,
      reason:
        'native backend: no executable aeon-monitor (set GNOSIS_AEON_MONITOR_BIN or build distributed-inference with --bin aeon-monitor)',
    };
  }
  return { ok: true, bin };
}

/**
 * @returns {{ ok: true } | { ok: false; reason: string }}
 */
export function probeServiceBackend() {
  const raw = process.env.MESH_STATION_URLS?.trim();
  if (!raw) {
    return {
      ok: false,
      reason:
        'service backend requires MESH_STATION_URLS (comma-separated fat-station base URLs for GET /health)',
    };
  }
  return { ok: true };
}

/**
 * @param {string} gnosisRoot
 * @param {'host' | 'native' | 'service'} backend
 * @returns {{ ok: true; nativeBin?: string } | { ok: false; reason: string }}
 */
export function preflightMonsterMesh(gnosisRoot, backend) {
  if (backend === 'host') {
    const p = probeHostBackend();
    if (!p.ok) return p;
    return { ok: true };
  }
  if (backend === 'native') {
    const n = probeNativeBackend(gnosisRoot);
    if (!n.ok) return n;
    return { ok: true, nativeBin: n.bin };
  }
  const s = probeServiceBackend();
  if (!s.ok) return s;
  return { ok: true };
}

const DELEGATION_SCHEMA_VERSION = 'monster-mesh.delegation.v1';

/**
 * @param {'host' | 'native' | 'service'} backend
 * @param {string | null} profile
 * @param {unknown} opaque
 */
export function formatDelegationEnvelope(backend, profile, opaque) {
  return JSON.stringify(
    {
      delegatedMonsterMesh: {
        schemaVersion: DELEGATION_SCHEMA_VERSION,
        backend,
        profile: profile ?? undefined,
        opaque,
      },
    },
    null,
    2,
  );
}

/**
 * @param {string[]} stationUrls
 * @returns {Promise<{ stations: { url: string; status: number; ok: boolean }[] }>}
 */
export async function fetchFatStationHealthRows(stationUrls) {
  const stations = [];
  for (const url of stationUrls) {
    const base = url.replace(/\/$/, '');
    let status = 0;
    let ok = false;
    try {
      const res = await fetch(`${base}/health`, {
        method: 'GET',
        signal: AbortSignal.timeout(10_000),
      });
      status = res.status;
      ok = res.ok;
    } catch {
      ok = false;
    }
    stations.push({ url: base, status, ok });
  }
  return { stations };
}

/**
 * @param {string} binDir path to open-source/gnosis/bin
 * @param {string[]} meshArgv full argv after node (includes monster-mesh...)
 */
export async function runMonsterMeshFromWrapper(binDir, meshArgv) {
  const gnosisRoot = resolve(binDir, '..');
  const sub = meshArgv[0];
  if (sub !== 'monster-mesh') {
    throw new Error('runMonsterMeshFromWrapper: expected monster-mesh');
  }
  const { rest, meshBackend: explicitBackend, profile } = parseMonsterMeshArgv(
    meshArgv.slice(1),
  );

  if (profile && !['ism433', 'ism915', 'nwr162'].includes(profile)) {
    console.error(
      `ERROR: invalid --profile "${profile}" (expected ism433|ism915|nwr162)`,
    );
    process.exit(2);
  }

  let backend;
  try {
    backend = resolveMonsterMeshBackend(explicitBackend, gnosisRoot);
  } catch (e) {
    console.error(`ERROR: ${(e).message}`);
    process.exit(2);
  }

  const pre = preflightMonsterMesh(gnosisRoot, backend);
  if (!pre.ok) {
    console.error(`REFUSED [monster-mesh backend=${backend}]: ${pre.reason}`);
    process.exit(2);
  }

  const profileEnv = profileEnvForDelegation(profile);
  const childEnv = { ...process.env, ...profileEnv };

  const require = createRequire(import.meta.url);
  const tsxImportPath = require.resolve('tsx');

  if (backend === 'host') {
    const meshEntry = resolve(
      gnosisRoot,
      'distributed-inference-host/src/bin/rknot-mesh-smoke.ts',
    );
    const mesh = spawnSync(
      process.execPath,
      ['--import', tsxImportPath, meshEntry, ...rest],
      { stdio: 'inherit', env: childEnv },
    );
    if (mesh.error) throw mesh.error;
    process.exit(mesh.status ?? 1);
  }

  if (backend === 'native') {
    const bin = pre.nativeBin;
    if (!bin) {
      console.error('REFUSED [monster-mesh backend=native]: internal error (missing binary path)');
      process.exit(2);
    }
    const mesh = spawnSync(bin, rest, {
      stdio: 'inherit',
      env: childEnv,
      cwd: gnosisRoot,
    });
    if (mesh.error) throw mesh.error;
    process.exit(mesh.status ?? 1);
  }

  // service: fat-station wire contract — GET /health only (opaque payload versioned inside opaque)
  const urls = (process.env.MESH_STATION_URLS ?? '')
    .split(',')
    .map((s) => s.trim())
    .filter(Boolean);
  const rows = await fetchFatStationHealthRows(urls);
  const allOk = rows.stations.every((s) => s.ok);
  if (!allOk) {
    console.error(
      `REFUSED [monster-mesh backend=service]: one or more /health probes failed`,
    );
    console.error(JSON.stringify(rows, null, 2));
    process.exit(1);
  }
  const opaque = {
    version: 'fat-station.health.v1',
    ...rows,
  };
  console.log(formatDelegationEnvelope('service', profile, opaque));
  process.exit(0);
}

export { DELEGATION_SCHEMA_VERSION };
