// @a0n/gnosis/src/betty/coarsen.ts
import type { Diagnostic, DiagnosticCode, GraphAST } from './compiler';
import type { StabilityReport, StabilityStateAssessment } from './stability';

export interface FiberPartition {
  fineToCoarse: Map<string, string>;
  coarseNodeIds: string[];
  fibers: Map<string, string[]>;
}

export interface CoarseNodeRate {
  coarseNodeId: string;
  aggregateArrival: number;
  aggregateService: number;
  drift: number;
}

export interface CoarseningLeanData {
  fineNodeIds: string[];
  coarseNodeIds: string[];
  quotientMapIndices: number[];
  arrivalRates: number[];
  serviceRates: number[];
  driftGap: number;
}

export interface CoarseningSynthesisResult {
  kind: 'success' | 'unstable' | 'invalid';
  partition: FiberPartition;
  coarseRates: CoarseNodeRate[];
  coarseDriftGap: number | null;
  unstableCoarseNodes: string[];
  diagnostics: Diagnostic[];
  leanData: CoarseningLeanData | null;
}

function round3(value: number): number {
  return Math.round(value * 1000) / 1000;
}

/**
 * Handles the gnosis extract Fiber Partition workflow.
 */
export function extractFiberPartition(ast: GraphAST): FiberPartition | null {
  const fineToCoarse = new Map<string, string>();
  let hasFiber = false;

  for (const [nodeId, node] of ast.nodes) {
    const fiber = node.properties.fiber;
    if (fiber) {
      fineToCoarse.set(nodeId, fiber);
      hasFiber = true;
    } else {
      fineToCoarse.set(nodeId, `__singleton_${nodeId}`);
    }
  }

  if (!hasFiber) {
    return null;
  }

  const fibers = new Map<string, string[]>();
  for (const [nodeId, coarseId] of fineToCoarse) {
    const members = fibers.get(coarseId) ?? [];
    members.push(nodeId);
    fibers.set(coarseId, members);
  }

  const coarseNodeIds = [...fibers.keys()];

  return { fineToCoarse, coarseNodeIds, fibers };
}

/**
 * Handles the gnosis validate Partition workflow.
 */
export function validatePartition(
  partition: FiberPartition,
  stability: StabilityReport
): Diagnostic[] {
  const diagnostics: Diagnostic[] = [];
  const assessedNodeIds = new Set(
    stability.stateAssessments.map((a) => a.nodeId)
  );

  for (const nodeId of assessedNodeIds) {
    if (!partition.fineToCoarse.has(nodeId)) {
      diagnostics.push({
        line: 0,
        column: 0,
        code: 'ERR_COARSENING_UNCOVERED_NODE' as DiagnosticCode,
        message: `State-assessed node '${nodeId}' has no fiber assignment`,
        severity: 'error',
      });
    }
  }

  for (const assessment of stability.stateAssessments) {
    const service = Number(assessment.service);
    if (
      assessment.service !== null &&
      (!Number.isFinite(service) || service <= 0)
    ) {
      if (assessment.service !== null && Number.isFinite(service)) {
        diagnostics.push({
          line: 0,
          column: 0,
          code: 'ERR_COARSENING_SERVICE_NOT_POSITIVE' as DiagnosticCode,
          message: `Node '${assessment.nodeId}' has non-positive service rate: ${assessment.service}`,
          severity: 'error',
        });
      }
    }
  }

  return diagnostics;
}

function getNumericRate(value: string | null): number | null {
  if (value === null) return null;
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : null;
}

/**
 * Handles the gnosis synthesize Coarsening workflow.
 */
export function synthesizeCoarsening(
  ast: GraphAST,
  stability: StabilityReport,
  partition: FiberPartition
): CoarseningSynthesisResult {
  const diagnostics: Diagnostic[] = [];
  const assessmentsByNodeId = new Map<string, StabilityStateAssessment>();
  for (const assessment of stability.stateAssessments) {
    assessmentsByNodeId.set(assessment.nodeId, assessment);
  }

  const coarseArrival = new Map<string, number>();
  const coarseService = new Map<string, number>();

  for (const coarseId of partition.coarseNodeIds) {
    coarseArrival.set(coarseId, 0);
    coarseService.set(coarseId, 0);
  }

  let hasSymbolic = false;

  for (const [nodeId, coarseId] of partition.fineToCoarse) {
    const assessment = assessmentsByNodeId.get(nodeId);
    if (!assessment) continue;

    const arrival = getNumericRate(assessment.arrival);
    const service = getNumericRate(assessment.service);

    if (arrival === null && assessment.arrival !== null) {
      hasSymbolic = true;
      diagnostics.push({
        line: 0,
        column: 0,
        code: 'ERR_COARSENING_SYMBOLIC_RATE' as DiagnosticCode,
        message: `Node '${nodeId}' has symbolic arrival rate '${assessment.arrival}' -- cannot aggregate numerically`,
        severity: 'warning',
      });
    }

    if (service === null && assessment.service !== null) {
      hasSymbolic = true;
      diagnostics.push({
        line: 0,
        column: 0,
        code: 'ERR_COARSENING_SYMBOLIC_RATE' as DiagnosticCode,
        message: `Node '${nodeId}' has symbolic service rate '${assessment.service}' -- cannot aggregate numerically`,
        severity: 'warning',
      });
    }

    if (arrival !== null) {
      coarseArrival.set(
        coarseId,
        round3((coarseArrival.get(coarseId) ?? 0) + arrival)
      );
    }
    if (service !== null) {
      coarseService.set(
        coarseId,
        round3((coarseService.get(coarseId) ?? 0) + service)
      );
    }
  }

  if (hasSymbolic) {
    return {
      kind: 'invalid',
      partition,
      coarseRates: [],
      coarseDriftGap: null,
      unstableCoarseNodes: [],
      diagnostics,
      leanData: null,
    };
  }

  const coarseRates: CoarseNodeRate[] = [];
  const unstableCoarseNodes: string[] = [];
  let minDriftGap = Infinity;

  for (const coarseId of partition.coarseNodeIds) {
    const agg_arrival = coarseArrival.get(coarseId) ?? 0;
    const agg_service = coarseService.get(coarseId) ?? 0;
    const drift = round3(agg_arrival - agg_service);

    coarseRates.push({
      coarseNodeId: coarseId,
      aggregateArrival: agg_arrival,
      aggregateService: agg_service,
      drift,
    });

    if (drift >= 0) {
      unstableCoarseNodes.push(coarseId);
      diagnostics.push({
        line: 0,
        column: 0,
        code: 'ERR_COARSENING_DRIFT_POSITIVE' as DiagnosticCode,
        message: `Coarse fiber '${coarseId}' has non-negative drift ${drift} (arrival=${agg_arrival}, service=${agg_service})`,
        severity: 'error',
      });
    } else {
      const gap = -drift;
      if (gap < minDriftGap) {
        minDriftGap = gap;
      }
    }
  }

  if (unstableCoarseNodes.length > 0) {
    return {
      kind: 'unstable',
      partition,
      coarseRates,
      coarseDriftGap: null,
      unstableCoarseNodes,
      diagnostics,
      leanData: null,
    };
  }

  const driftGap = round3(minDriftGap);

  const fineNodeIds = [...ast.nodes.keys()];
  const coarseIndexMap = new Map<string, number>();
  for (let i = 0; i < partition.coarseNodeIds.length; i++) {
    coarseIndexMap.set(partition.coarseNodeIds[i]!, i);
  }

  const quotientMapIndices = fineNodeIds.map((nodeId) => {
    const coarseId = partition.fineToCoarse.get(nodeId)!;
    return coarseIndexMap.get(coarseId) ?? 0;
  });

  const arrivalRates = fineNodeIds.map((nodeId) => {
    const assessment = assessmentsByNodeId.get(nodeId);
    return assessment ? getNumericRate(assessment.arrival) ?? 0 : 0;
  });

  const serviceRates = fineNodeIds.map((nodeId) => {
    const assessment = assessmentsByNodeId.get(nodeId);
    return assessment ? getNumericRate(assessment.service) ?? 0 : 0;
  });

  return {
    kind: 'success',
    partition,
    coarseRates,
    coarseDriftGap: driftGap,
    unstableCoarseNodes: [],
    diagnostics,
    leanData: {
      fineNodeIds,
      coarseNodeIds: partition.coarseNodeIds,
      quotientMapIndices,
      arrivalRates,
      serviceRates,
      driftGap,
    },
  };
}
