// @a0n/gnosis/src/betty/gg-grassmannian-cache.ts
import { BettyCompiler, type ASTEdge, type ASTNode, type GraphAST } from './compiler';
import {
  MONSTER_GRASSMANNIAN_LANE_COUNT,
  type PackedGrassmannianLaneMask,
  type TopologicalXorCertificate,
  createPackedGrassmannianLaneMask,
  issueGrassmannianXorRoundCertificateFromMask,
  packedGrassmannianLaneMaskDefectRank,
  togglePackedGrassmannianLane,
} from '../topological-xor-certificate';

export { MONSTER_GRASSMANNIAN_LANE_COUNT };

export interface QBracketRuntime {
  readonly lower: number;
  readonly upper: number;
  readonly theoremRef: string;
}

export type GgGrassmannianCacheLaneKind =
  | 'node'
  | 'edge'
  | 'fork'
  | 'fold'
  | 'race'
  | 'process';

export interface GgGrassmannianCacheLane {
  readonly key: string;
  readonly laneIndex: number;
  readonly kind: GgGrassmannianCacheLaneKind;
  readonly fingerprint: string;
  readonly qBracket: QBracketRuntime;
  readonly label: string;
}

export interface GgGrassmannianCachePlan {
  readonly chart: string;
  readonly laneCount: number;
  readonly topologyFingerprint: string;
  readonly mask: PackedGrassmannianLaneMask;
  readonly lanes: readonly GgGrassmannianCacheLane[];
  readonly certificate: TopologicalXorCertificate;
}

export interface GgGrassmannianCacheDiff {
  readonly chart: string;
  readonly laneCount: number;
  readonly changedMask: PackedGrassmannianLaneMask;
  readonly changedLanes: readonly GgGrassmannianCacheLane[];
  readonly removedLanes: readonly GgGrassmannianCacheLane[];
  readonly defectRank: number;
  readonly certificate: TopologicalXorCertificate;
}

export type GgGrassmannianCacheSkipAction =
  | 'skip-current-execution'
  | 'replay-changed-lanes'
  | 'rebuild-app-cache';

export interface GgGrassmannianCacheSkipDecision {
  readonly action: GgGrassmannianCacheSkipAction;
  readonly canSkip: boolean;
  readonly reason: string;
  readonly defectRank: number;
  readonly diff?: GgGrassmannianCacheDiff;
}

export interface CompileGgGrassmannianCacheOptions {
  readonly chart?: string;
  readonly laneCount?: number;
  readonly sourceId?: string;
}

const QBRACKET_THEOREM_REF = 'Gnosis.BracketedSpace.QBracket';
const FANOUT_THEOREM_REF = 'Gnosis.BracketedSpace.phi_refines_0_1';

/**
 * Handles the gnosis compile Gg Topology To Grassmannian Cache workflow.
 */
export function compileGgTopologyToGrassmannianCache(
  source: string | GraphAST,
  options: CompileGgGrassmannianCacheOptions = {}
): GgGrassmannianCachePlan {
  const chart = options.chart ?? 'gg-topology-cache';
  const laneCount = options.laneCount ?? MONSTER_GRASSMANNIAN_LANE_COUNT;
  const ast = typeof source === 'string' ? parseGgSource(source) : source;
  const lanes = collectGrassmannianCacheLanes(ast, laneCount);
  const mask = createPackedGrassmannianLaneMask(laneCount);

  for (const lane of lanes) {
    togglePackedGrassmannianLane(mask, lane.laneIndex);
  }

  const topologyFingerprint = stableHash64(
    lanes.map((lane) => `${lane.key}:${lane.fingerprint}:${lane.laneIndex}`).join('\n')
  );
  const certificate = issueGrassmannianXorRoundCertificateFromMask({
    roundId: options.sourceId ?? topologyFingerprint,
    chart,
    mask,
    theoremRefs: [QBRACKET_THEOREM_REF, FANOUT_THEOREM_REF],
  });

  return {
    chart,
    laneCount,
    topologyFingerprint,
    mask,
    lanes,
    certificate,
  };
}

/**
 * Handles the gnosis decide Gg Grassmannian Cache Skip workflow.
 */
export function decideGgGrassmannianCacheSkip(
  seed: GgGrassmannianCachePlan,
  current: GgGrassmannianCachePlan
): GgGrassmannianCacheSkipDecision {
  if (seed.laneCount !== current.laneCount || seed.chart !== current.chart) {
    return {
      action: 'rebuild-app-cache',
      canSkip: false,
      reason: 'Grassmannian chart changed; rebuild the app cache before replay decisions',
      defectRank: current.certificate.defectRank,
    };
  }

  if (seed.topologyFingerprint === current.topologyFingerprint) {
    return {
      action: 'skip-current-execution',
      canSkip: true,
      reason: 'seed and current execution fingerprints match',
      defectRank: 0,
    };
  }

  const diff = diffGgGrassmannianCachePlans(seed, current);
  if (diff.defectRank === 0) {
    return {
      action: 'skip-current-execution',
      canSkip: true,
      reason: 'fingerprint changed but certified lane diff is balanced',
      defectRank: 0,
      diff,
    };
  }

  return {
    action: 'replay-changed-lanes',
    canSkip: false,
    reason: 'current execution changed sparse Grassmannian lanes',
    defectRank: diff.defectRank,
    diff,
  };
}

/**
 * Handles the gnosis diff Gg Grassmannian Cache Plans workflow.
 */
export function diffGgGrassmannianCachePlans(
  previous: GgGrassmannianCachePlan,
  next: GgGrassmannianCachePlan
): GgGrassmannianCacheDiff {
  if (previous.laneCount !== next.laneCount) {
    throw new Error('GG Grassmannian cache plans must use the same lane count');
  }
  if (previous.chart !== next.chart) {
    throw new Error('GG Grassmannian cache plans must use the same chart');
  }

  const previousByKey = new Map(previous.lanes.map((lane) => [lane.key, lane]));
  const nextByKey = new Map(next.lanes.map((lane) => [lane.key, lane]));
  const changedMask = createPackedGrassmannianLaneMask(next.laneCount);
  const changedLanes: GgGrassmannianCacheLane[] = [];
  const removedLanes: GgGrassmannianCacheLane[] = [];

  for (const lane of next.lanes) {
    const previousLane = previousByKey.get(lane.key);
    if (!previousLane || previousLane.fingerprint !== lane.fingerprint) {
      togglePackedGrassmannianLane(changedMask, lane.laneIndex);
      changedLanes.push(lane);
    }
  }

  for (const lane of previous.lanes) {
    if (!nextByKey.has(lane.key)) {
      togglePackedGrassmannianLane(changedMask, lane.laneIndex);
      removedLanes.push(lane);
    }
  }

  const certificate = issueGrassmannianXorRoundCertificateFromMask({
    roundId: `${previous.topologyFingerprint}->${next.topologyFingerprint}`,
    chart: next.chart,
    mask: changedMask,
    theoremRefs: [QBRACKET_THEOREM_REF, FANOUT_THEOREM_REF],
  });

  return {
    chart: next.chart,
    laneCount: next.laneCount,
    changedMask,
    changedLanes: changedLanes.sort(compareLane),
    removedLanes: removedLanes.sort(compareLane),
    defectRank: packedGrassmannianLaneMaskDefectRank(changedMask),
    certificate,
  };
}

function parseGgSource(source: string): GraphAST {
  const result = new BettyCompiler().parse(source);
  const errors = result.diagnostics.filter((diagnostic) => diagnostic.severity === 'error');
  if (errors.length > 0) {
    throw new Error(`GG topology parse failed: ${errors.map((d) => d.message).join('; ')}`);
  }
  return result.ast ?? { nodes: new Map(), edges: [] };
}

function collectGrassmannianCacheLanes(
  ast: GraphAST,
  laneCount: number
): GgGrassmannianCacheLane[] {
  const reserved = new Set<number>();
  const lanes: GgGrassmannianCacheLane[] = [];

  const nodes = Array.from(ast.nodes.values()).sort((a, b) => a.id.localeCompare(b.id));
  for (const node of nodes) {
    lanes.push(makeNodeLane(node, laneCount, reserved));
  }

  const edges = ast.edges.slice().sort(compareEdges);
  for (let edgeOrdinal = 0; edgeOrdinal < edges.length; edgeOrdinal++) {
    lanes.push(makeEdgeLane(edges[edgeOrdinal]!, edgeOrdinal, laneCount, reserved));
  }

  return lanes.sort(compareLane);
}

function makeNodeLane(
  node: ASTNode,
  laneCount: number,
  reserved: Set<number>
): GgGrassmannianCacheLane {
  const canonical = canonicalNode(node);
  return makeLane({
    key: `node:${node.id}`,
    kind: 'node',
    label: node.labels[0] ?? 'Node',
    canonical,
    bracketWidth: 1,
    laneCount,
    reserved,
  });
}

function makeEdgeLane(
  edge: ASTEdge,
  edgeOrdinal: number,
  laneCount: number,
  reserved: Set<number>
): GgGrassmannianCacheLane {
  const edgeType = edge.type.toUpperCase();
  const kind = edgeKind(edgeType);
  const canonical = canonicalEdge(edge, edgeOrdinal);
  return makeLane({
    key: `edge:${canonical}`,
    kind,
    label: edgeType,
    canonical,
    bracketWidth: Math.max(1, edge.sourceIds.length * edge.targetIds.length),
    laneCount,
    reserved,
  });
}

function makeLane(input: {
  readonly key: string;
  readonly kind: GgGrassmannianCacheLaneKind;
  readonly label: string;
  readonly canonical: string;
  readonly bracketWidth: number;
  readonly laneCount: number;
  readonly reserved: Set<number>;
}): GgGrassmannianCacheLane {
  const fingerprint = stableHash64(input.canonical);
  const laneIndex = allocateLane(input.key, input.laneCount, input.reserved);
  return {
    key: input.key,
    kind: input.kind,
    label: input.label,
    fingerprint,
    laneIndex,
    qBracket: {
      lower: laneIndex,
      upper: laneIndex + input.bracketWidth,
      theoremRef: QBRACKET_THEOREM_REF,
    },
  };
}

function allocateLane(key: string, laneCount: number, reserved: Set<number>): number {
  if (!Number.isSafeInteger(laneCount) || laneCount <= 0) {
    throw new Error('laneCount must be a positive safe integer');
  }
  let laneIndex = Number(hash64(key) % BigInt(laneCount));
  for (let probes = 0; probes < laneCount; probes++) {
    if (!reserved.has(laneIndex)) {
      reserved.add(laneIndex);
      return laneIndex;
    }
    laneIndex = (laneIndex + 1) % laneCount;
  }
  throw new Error('GG topology has more cache lanes than the Grassmannian chart can hold');
}

function edgeKind(edgeType: string): GgGrassmannianCacheLaneKind {
  switch (edgeType) {
    case 'FORK':
      return 'fork';
    case 'FOLD':
      return 'fold';
    case 'RACE':
      return 'race';
    case 'PROCESS':
      return 'process';
    default:
      return 'edge';
  }
}

function canonicalNode(node: ASTNode): string {
  return stableJson({
    id: node.id,
    labels: node.labels,
    properties: node.properties,
  });
}

function canonicalEdge(edge: ASTEdge, edgeOrdinal: number): string {
  return stableJson({
    edgeOrdinal,
    sourceIds: edge.sourceIds,
    targetIds: edge.targetIds,
    type: edge.type,
    properties: edge.properties,
  });
}

function compareEdges(left: ASTEdge, right: ASTEdge): number {
  return canonicalEdge(left, 0).localeCompare(canonicalEdge(right, 0));
}

function compareLane(left: GgGrassmannianCacheLane, right: GgGrassmannianCacheLane): number {
  return left.laneIndex - right.laneIndex || left.key.localeCompare(right.key);
}

function stableJson(value: unknown): string {
  if (value === null || typeof value !== 'object') {
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) {
    return `[${value.map(stableJson).join(',')}]`;
  }
  const record = value as Record<string, unknown>;
  return `{${Object.keys(record)
    .sort()
    .map((key) => `${JSON.stringify(key)}:${stableJson(record[key])}`)
    .join(',')}}`;
}

function stableHash64(value: string): string {
  return hash64(value).toString(16).padStart(16, '0');
}

function hash64(value: string): bigint {
  let hash = 0xcbf29ce484222325n;
  for (let i = 0; i < value.length; i++) {
    hash ^= BigInt(value.charCodeAt(i));
    hash = BigInt.asUintN(64, hash * 0x100000001b3n);
  }
  return hash;
}
