// @a0n/gnosis/src/betty/hologram-navigator.ts
/**
 * Hologram Navigator: The Topological Raycaster
 *
 * Provides O(1) mathematical projections into Grassmannian Tensor Holograms,
 * entirely bypassing string parsing and sequential search.
 */

import type { HologramPayload } from './data-hologram';

interface RaycasterWasmModule {
  project_affine_frf(
    hologramJson: string,
    row: number,
    col: number
  ): unknown;
  intersect_schubert_frf?(
    hologramJson: string,
    col: number,
    op: string,
    val: string
  ): unknown;
}

let RaycasterWasm: RaycasterWasmModule | null = null;
let raycasterLoadStarted = false;
const RAYCASTER_WASM_MODULE_PATH: string =
  '../../../knotgraph/rust/crates/knotgraph-raycaster/pkg/knotgraph_raycaster.js';

function isRaycasterWasmModule(value: unknown): value is RaycasterWasmModule {
  return (
    typeof value === 'object' &&
    value !== null &&
    'project_affine_frf' in value &&
    typeof (value as { project_affine_frf?: unknown }).project_affine_frf ===
      'function'
  );
}

function startRaycasterLoad(): void {
  if (
    raycasterLoadStarted ||
    typeof process === 'undefined' ||
    process.release?.name !== 'node'
  ) {
    return;
  }

  raycasterLoadStarted = true;
  void import(RAYCASTER_WASM_MODULE_PATH)
    .then((module: unknown) => {
      RaycasterWasm = isRaycasterWasmModule(module) ? module : null;
    })
    .catch(() => {
      RaycasterWasm = null;
    });
}

export class HologramNavigator {
  private hologram: HologramPayload;

  constructor(hologram: HologramPayload) {
    this.hologram = hologram;
  }

  /**
   * Affine Coordinate Projection
   * Computes O(1) mathematical lookup into a matrix (e.g. CSV arrays).
   */
  public projectAffine(row: number, col: number): string {
    startRaycasterLoad();
    try {
      // 1. Attempt High-Performance Rust SIMD/WASM Raycast
      if (RaycasterWasm) {
        const hologramJson = JSON.stringify(this.hologram);
        return RaycasterWasm.project_affine_frf(
          hologramJson,
          row,
          col
        ) as string;
      }
    } catch (e) {
      // Fallback to TS implementation if WASM fails to load or throws
    }

    const { dimensionality: k, staticVolume: V } = this.hologram.geometry;

    const targetIndex = row * k + col;

    // Mathematically verify the coordinate is within the positive Grassmannian boundary
    if (targetIndex < 0 || targetIndex >= V) {
      throw new Error(
        `Out of bounds: Coordinate [${row}, ${col}] exceeds static volume ${V}.`
      );
    }

    if (col >= k || col < 0) {
      throw new Error(
        `Out of bounds: Column ${col} exceeds dimensionality ${k}.`
      );
    }

    return this.hologram.ventDictionary[targetIndex]!;
  }

  /**
   * Boundary Operator Traversal
   * Navigates sequence topology dimensions (e.g. Markdown block structures)
   * returning a sliced geometric sub-hologram.
   */
  public traverseBoundary(
    predicate: (nodeText: string, index: number) => boolean
  ): HologramPayload {
    const slicedDictionary: string[] = [];
    const { ventDictionary, geometry } = this.hologram;

    for (let i = 0; i < geometry.staticVolume; i++) {
      const text = ventDictionary[i]!;
      if (predicate(text, i)) {
        slicedDictionary.push(text);
      }
    }

    return {
      geometry: {
        dimensionality: geometry.dimensionality, // k remains intact (schema is unchanged)
        staticVolume: slicedDictionary.length, // n alters
        originVertex: geometry.originVertex,
        flattenedFaces: 0, // derived subsets clear boundary defects
      },
      ventDictionary: slicedDictionary,
    };
  }

  /**
   * Schubert Slicing (Semantic Value Queries)
   * Mathematically intersects requested indices back through the Gr(k,n) boundary tensor.
   * Emits a smaller, structurally identical Hologram representing the subset.
   */
  public intersectSchubert(
    predicate: (val: string, colIndex: number) => boolean,
    wasmFilters?: { col: number; op: string; val: string }
  ): HologramPayload {
    startRaycasterLoad();
    // 1. Attempt High-Performance Rust SIMD/WASM Schubert Intersection
    if (wasmFilters && RaycasterWasm?.intersect_schubert_frf) {
      try {
        const hologramJson = JSON.stringify(this.hologram);
        const resultJson = RaycasterWasm.intersect_schubert_frf(
          hologramJson,
          wasmFilters.col,
          wasmFilters.op,
          wasmFilters.val
        ) as string;

        return JSON.parse(resultJson) as HologramPayload;
      } catch (e) {
        console.error('[WASM Fallback] Error in Raycaster WASM:', e);
        // Fallback to TS
      }
    }

    const { dimensionality: k, staticVolume: V } = this.hologram.geometry;
    const { ventDictionary } = this.hologram;
    const n = V / k;

    const subsetDictionary: string[] = [];

    // Iterate the n constraint rows
    for (let row = 0; row < n; row++) {
      let includeRow = false;
      const rowOffset = row * k;

      // Check Schubert Intersection
      for (let col = 0; col < k; col++) {
        const val = ventDictionary[rowOffset + col]!;
        if (predicate(val, col)) {
          includeRow = true;
          break; // Schubert cell bounds isolated
        }
      }

      // If the row intersects the query cell, project it into the new subset tensor
      if (includeRow) {
        for (let col = 0; col < k; col++) {
          subsetDictionary.push(ventDictionary[rowOffset + col]!);
        }
      }
    }

    return {
      geometry: {
        dimensionality: k,
        staticVolume: subsetDictionary.length, // updated n * k
        originVertex: this.hologram.geometry.originVertex,
        flattenedFaces: 0,
      },
      ventDictionary: subsetDictionary,
    };
  }
}
