// @a0n/gnosis/src/betty/hopf-algebra.ts
/**
 * Connes-Kreimer Hopf Algebra Substrate
 * 
 * Maps algorithmic execution graphs (ASTs) to the Cosmic Galois Group.
 * In a Hopf algebra, every divergent loop structure can be mathematically
 * factored into an infinite pole and a finite residue.
 */

export interface ASTNode {
  id: string;
  isInfiniteCycle: boolean;
  payload: any;
  children: ASTNode[];
}

export interface Coproduct {
  subforest: ASTNode[];  // The divergent structures (T_-)
  quotient: ASTNode | null; // The stable algebraic residue (T_+)
}

export class HopfAlgebra {
  /**
   * The Coproduct (Δ)
   * Splits a topological execution graph into divergent cycles and stable residues.
   * If a node is an infinite cycle, it is sheared off into the subforest.
   */
  public static coproduct(tree: ASTNode): Coproduct {
    const subforest: ASTNode[] = [];
    let quotient: ASTNode | null = null;

    if (tree.isInfiniteCycle) {
      // The root itself is a divergent infinity.
      subforest.push(tree);
      // The quotient is mathematically the empty tree (unit), 
      // but the residue data is trapped inside the cycle's payload.
      // We extract the finite algebraic truth trapped inside the infinity.
      quotient = {
        id: `Residue(${tree.id})`,
        isInfiniteCycle: false,
        payload: tree.payload,
        children: []
      };
    } else {
      // It is stable, so it forms the quotient. 
      // We must recursively apply the coproduct to its children.
      const safeChildren: ASTNode[] = [];
      for (const child of tree.children) {
        const childDelta = this.coproduct(child);
        subforest.push(...childDelta.subforest);
        if (childDelta.quotient) {
          safeChildren.push(childDelta.quotient);
        }
      }
      
      quotient = {
        id: tree.id,
        isInfiniteCycle: false,
        payload: tree.payload,
        children: safeChildren
      };
    }

    return { subforest, quotient };
  }

  /**
   * The Antipode (S)
   * Algebraically inverts the divergent subforest, allowing it to be mathematically
   * subtracted from the execution timeline during Birkhoff factorization.
   */
  public static antipode(subforest: ASTNode[]): ASTNode[] {
    return subforest.map(node => ({
      ...node,
      // The topological inverse of an infinite pole is a Void sink.
      id: `InversePole(${node.id})`,
      payload: null 
    }));
  }
}
