// @a0n/gnosis/src/betty/optimizer.ts
/**
 * Theorem-backed optimization pass manager for the Betty compiler.
 *
 * Each pass is backed by a mechanized theorem from the formal ledger.
 * The pass manager runs passes in priority order, accumulating diagnostics
 * and proof obligations that feed into lean.ts certificate emission.
 */

import type {
  ASTEdge,
  ASTNode,
  Diagnostic,
  DiagnosticCode,
  GraphAST,
} from './compiler';
import type { StabilityReport } from './stability';
import {
  classifyFoldRegime,
  idleStages,
  type FoldRegime,
} from '../reynolds-bft';
import { buleyeanWeight } from '../void';

// ─── Pass result ─────────────────────────────────────────────────────

export interface OptimizationPassResult {
  ast: GraphAST;
  diagnostics: Diagnostic[];
  certificates: OptimizationCertificate[];
  applied: boolean;
}

export interface OptimizationCertificate {
  passName: string;
  theoremId: string;
  leanTheoremName: string;
  summary: string;
  data: Record<string, unknown>;
}

// ─── Pass interface ──────────────────────────────────────────────────

export interface OptimizationPass {
  readonly name: string;
  readonly theoremId: string;
  readonly priority: number;
  /** Whether this pass modifies the AST ('transform') or only analyzes ('analyze').
   *  Analysis-only passes are FORKED in parallel per THM-SERVER-OPTIMALITY. */
  readonly kind: 'transform' | 'analyze';
  predicate(ast: GraphAST, stability: StabilityReport): boolean;
  apply(ast: GraphAST, stability: StabilityReport): OptimizationPassResult;
}

interface FoldSegmentMetrics {
  stages: number;
  chunks: number;
  idle: number;
  regime: FoldRegime;
}

function isMultiSourceFold(edge: ASTEdge): boolean {
  return (
    (edge.type === 'FOLD' || edge.type === 'COLLAPSE') &&
    edge.sourceIds.length > 1
  );
}

function dedupeIds(ids: string[]): string[] {
  return [...new Set(ids)];
}

function buildProcessPredecessors(ast: GraphAST): Map<string, string[]> {
  const predecessors = new Map<string, string[]>();

  for (const edge of ast.edges) {
    if (edge.type !== 'PROCESS') {
      continue;
    }

    for (const targetId of edge.targetIds) {
      const current = predecessors.get(targetId) ?? [];
      current.push(...edge.sourceIds);
      predecessors.set(targetId, dedupeIds(current));
    }
  }

  return predecessors;
}

function inferFoldSegmentStages(ast: GraphAST, sourceIds: string[]): number {
  const predecessors = buildProcessPredecessors(ast);
  let frontier = dedupeIds(sourceIds);
  const seen = new Set(frontier);
  let stages = 0;

  while (frontier.length > 0) {
    stages += 1;

    const nextLayer = new Set<string>();
    for (const nodeId of frontier) {
      const nodePredecessors = predecessors.get(nodeId) ?? [];
      for (const predecessorId of nodePredecessors) {
        if (seen.has(predecessorId)) {
          continue;
        }
        seen.add(predecessorId);
        nextLayer.add(predecessorId);
      }
    }

    frontier = [...nextLayer];
  }

  return Math.max(stages, 1);
}

function inferFoldSegmentMetrics(
  ast: GraphAST,
  edge: ASTEdge
): FoldSegmentMetrics {
  const stages = inferFoldSegmentStages(ast, edge.sourceIds);
  const chunks = edge.sourceIds.length;
  return {
    stages,
    chunks,
    idle: idleStages(stages, chunks),
    regime: classifyFoldRegime(stages, chunks),
  };
}

function theoremNameForFoldRegime(regime: FoldRegime): string {
  switch (regime) {
    case 'mergeAll':
      return 'classifyRegime_eq_mergeAll_iff_quorumSafe';
    case 'quorumFold':
      return 'classifyRegime_eq_quorumFold_iff_majoritySafe_not_quorumSafe';
    case 'syncRequired':
      return 'classifyRegime_eq_syncRequired_iff_not_majoritySafe';
  }
}

function minimumMajorityChunks(stages: number): number {
  return Math.floor(stages / 2) + 1;
}

function collectProcessAncestors(
  ast: GraphAST,
  nodeIds: string[]
): Set<string> {
  const predecessors = buildProcessPredecessors(ast);
  const pending = [...nodeIds];
  const visited = new Set(nodeIds);

  while (pending.length > 0) {
    const current = pending.pop()!;
    const currentPredecessors = predecessors.get(current) ?? [];
    for (const predecessorId of currentPredecessors) {
      if (visited.has(predecessorId)) {
        continue;
      }
      visited.add(predecessorId);
      pending.push(predecessorId);
    }
  }

  return visited;
}

function canRemoveProcessLineageNode(
  ast: GraphAST,
  nodeId: string,
  removableNodes: Set<string>,
  foldEdgeIndex: number
): boolean {
  for (let index = 0; index < ast.edges.length; index += 1) {
    const edge = ast.edges[index];
    if (!edge) continue;
    if (!edge.sourceIds.includes(nodeId)) {
      continue;
    }

    if (index === foldEdgeIndex) {
      continue;
    }

    const liveTargets = edge.targetIds.filter((targetId) => {
      return !removableNodes.has(targetId);
    });

    if (liveTargets.length > 0) {
      return false;
    }
  }

  return true;
}

function pruneFoldLineage(
  ast: GraphAST,
  foldEdgeIndex: number,
  keptSourceIds: string[]
): GraphAST {
  const foldEdge = ast.edges[foldEdgeIndex];
  if (!foldEdge) return ast;
  const keptSourceSet = new Set(keptSourceIds);
  const droppedSourceIds = foldEdge.sourceIds.filter((sourceId) => {
    return !keptSourceSet.has(sourceId);
  });

  if (droppedSourceIds.length === 0) {
    return ast;
  }

  const candidateNodes = collectProcessAncestors(ast, droppedSourceIds);
  const removableNodes = new Set(candidateNodes);
  let changed = true;

  while (changed) {
    changed = false;
    for (const nodeId of [...removableNodes]) {
      if (
        !canRemoveProcessLineageNode(ast, nodeId, removableNodes, foldEdgeIndex)
      ) {
        removableNodes.delete(nodeId);
        changed = true;
      }
    }
  }

  const nextNodes = new Map<string, ASTNode>();
  for (const [nodeId, node] of ast.nodes) {
    if (!removableNodes.has(nodeId)) {
      nextNodes.set(nodeId, node);
    }
  }

  const nextEdges: ASTEdge[] = [];
  for (let index = 0; index < ast.edges.length; index += 1) {
    const edge = ast.edges[index];
    if (!edge) continue;
    const sourceIds =
      index === foldEdgeIndex
        ? edge.sourceIds.filter((sourceId) => keptSourceSet.has(sourceId))
        : edge.sourceIds.filter((sourceId) => !removableNodes.has(sourceId));
    const targetIds = edge.targetIds.filter((targetId) => {
      return !removableNodes.has(targetId);
    });

    if (sourceIds.length === 0 || targetIds.length === 0) {
      continue;
    }

    nextEdges.push({ ...edge, sourceIds, targetIds } as ASTEdge);
  }

  return { nodes: nextNodes, edges: nextEdges };
}

export class ReynoldsFoldRegimePass implements OptimizationPass {
  readonly name = 'reynolds-fold-regime';
  readonly theoremId = 'THM-REYNOLDS-BFT-REGIME';
  readonly priority = 5;
  readonly kind = 'transform' as const;

  predicate(ast: GraphAST): boolean {
    return ast.edges.some(isMultiSourceFold);
  }

  apply(ast: GraphAST): OptimizationPassResult {
    let currentAst = ast;
    const diagnostics: Diagnostic[] = [];
    const certificates: OptimizationCertificate[] = [];
    const candidates = ast.edges.filter(isMultiSourceFold).map((edge) => ({
      type: edge.type,
      targetKey: edge.targetIds.join('|'),
      sourceIds: [...edge.sourceIds],
    }));

    for (const candidate of candidates) {
      const foldEdgeIndex = currentAst.edges.findIndex((edge) => {
        return (
          edge.type === candidate.type &&
          edge.targetIds.join('|') === candidate.targetKey &&
          candidate.sourceIds.some((sourceId) =>
            edge.sourceIds.includes(sourceId)
          )
        );
      });
      if (foldEdgeIndex === -1) {
        continue;
      }

      const edge = currentAst.edges[foldEdgeIndex];
      if (!edge || !isMultiSourceFold(edge)) {
        continue;
      }

      const metrics = inferFoldSegmentMetrics(currentAst, edge);
      const theoremName = theoremNameForFoldRegime(metrics.regime);
      const baseData = {
        edgeType: edge.type,
        sourceIds: [...edge.sourceIds],
        targetIds: [...edge.targetIds],
        stages: metrics.stages,
        chunks: metrics.chunks,
        idleStages: metrics.idle,
        regime: metrics.regime,
      };

      if (metrics.regime === 'quorumFold') {
        const keptChunkCount = minimumMajorityChunks(metrics.stages);
        const keptSourceIds = edge.sourceIds.slice(0, keptChunkCount);
        const droppedSourceIds = edge.sourceIds.slice(keptChunkCount);
        const prunedAst = pruneFoldLineage(
          currentAst,
          foldEdgeIndex,
          keptSourceIds
        );
        const nextFoldIndex = prunedAst.edges.findIndex((candidateEdge) => {
          return (
            candidateEdge.targetIds.join('|') === edge.targetIds.join('|') &&
            candidateEdge.sourceIds.join('|') === keptSourceIds.join('|') &&
            candidateEdge.type === edge.type
          );
        });
        const nextEdges = [...prunedAst.edges];
        if (nextFoldIndex !== -1) {
          const nextEdge = nextEdges[nextFoldIndex]!;
          nextEdges[nextFoldIndex] = {
            ...nextEdge,
            properties: {
              ...nextEdge.properties,
              reynoldsRegime: metrics.regime,
              reynoldsStages: metrics.stages.toString(),
              reynoldsChunks: metrics.chunks.toString(),
              reynoldsIdleStages: metrics.idle.toString(),
              reynoldsKeptChunks: keptSourceIds.length.toString(),
              reynoldsTheorem: theoremName,
            },
          };
        }
        currentAst = { nodes: prunedAst.nodes, edges: nextEdges };

        diagnostics.push({
          line: 1,
          column: 1,
          message: `[${this.theoremId}] Fold [${edge.sourceIds.join(
            ', '
          )}]→[${edge.targetIds.join(
            ', '
          )}] classified quorumFold with stages=${metrics.stages}, chunks=${
            metrics.chunks
          }. Pruned to deterministic majority subset [${keptSourceIds.join(
            ', '
          )}] by ${theoremName}.`,
          severity: 'info',
        });
        certificates.push({
          passName: this.name,
          theoremId: this.theoremId,
          leanTheoremName: theoremName,
          summary: `Fold [${edge.sourceIds.join(', ')}]→[${edge.targetIds.join(
            ', '
          )}] classified quorumFold and pruned ${
            droppedSourceIds.length
          } branch(es) before coarsening.`,
          data: {
            ...baseData,
            keptSourceIds,
            droppedSourceIds,
          },
        });
        continue;
      }

      const nextEdges = [...currentAst.edges];
      const nextEdge = nextEdges[foldEdgeIndex]!;
      nextEdges[foldEdgeIndex] = {
        ...nextEdge,
        type:
          metrics.regime === 'syncRequired' && nextEdge.type === 'COLLAPSE'
            ? 'FOLD'
            : nextEdge.type,
        properties: {
          ...nextEdge.properties,
          reynoldsRegime: metrics.regime,
          reynoldsStages: metrics.stages.toString(),
          reynoldsChunks: metrics.chunks.toString(),
          reynoldsIdleStages: metrics.idle.toString(),
          reynoldsTheorem: theoremName,
        },
      };
      currentAst = { nodes: currentAst.nodes, edges: nextEdges };

      diagnostics.push({
        line: 1,
        column: 1,
        message:
          metrics.regime === 'syncRequired'
            ? `[${this.theoremId}] Fold [${edge.sourceIds.join(
                ', '
              )}]→[${edge.targetIds.join(
                ', '
              )}] classified syncRequired with stages=${
                metrics.stages
              }, chunks=${
                metrics.chunks
              }. Automatic collapse is blocked by ${theoremName}.`
            : `[${this.theoremId}] Fold [${edge.sourceIds.join(
                ', '
              )}]→[${edge.targetIds.join(
                ', '
              )}] classified mergeAll with stages=${metrics.stages}, chunks=${
                metrics.chunks
              }. Full coarsening remains admissible by ${theoremName}.`,
        severity: metrics.regime === 'syncRequired' ? 'warning' : 'info',
      });
      certificates.push({
        passName: this.name,
        theoremId: this.theoremId,
        leanTheoremName: theoremName,
        summary:
          metrics.regime === 'syncRequired'
            ? `Fold [${edge.sourceIds.join(', ')}]→[${edge.targetIds.join(
                ', '
              )}] classified syncRequired and blocked automatic collapse.`
            : `Fold [${edge.sourceIds.join(', ')}]→[${edge.targetIds.join(
                ', '
              )}] classified mergeAll and remained fully coarsenable.`,
        data: baseData,
      });
    }

    return {
      ast: currentAst,
      diagnostics,
      certificates,
      applied: certificates.length > 0,
    };
  }
}

// ─── Coarsening pass (THM-RECURSIVE-COARSENING-SYNTHESIS) ────────────
//
// Backed by RecursiveCoarseningSynthesis.lean:
//   synthesis_sound: if all coarse nodes have negative drift, certificate is valid
//   drift_conservation: total fine drift = total coarse drift
//   fine_stability_implies_coarse_stability: stable fine graph → stable coarse graph
//
// The pass identifies subgraphs that can be collapsed into coarse nodes
// while preserving stability certificates.

export interface CoarseNodeData {
  coarseId: string;
  fineNodeIds: string[];
  aggregateArrival: number;
  aggregateService: number;
  coarseDrift: number;
}

export class CoarseningPass implements OptimizationPass {
  readonly name = 'coarsening';
  readonly theoremId = 'THM-RECURSIVE-COARSENING-SYNTHESIS';
  readonly priority = 10;
  readonly kind = 'transform' as const;

  predicate(ast: GraphAST, stability: StabilityReport): boolean {
    // Only coarsen when we have stability data and enough nodes to coarsen
    if (!stability.enabled || ast.nodes.size < 4) {
      return false;
    }
    // Need at least some stable states to synthesize quotient
    return stability.stateAssessments.some((sa) => sa.status === 'stable');
  }

  apply(ast: GraphAST, stability: StabilityReport): OptimizationPassResult {
    const diagnostics: Diagnostic[] = [];
    const certificates: OptimizationCertificate[] = [];

    // Identify candidate fiber groups: nodes with same drift sign that
    // share only internal edges between them
    const groups = this.identifyCoarseningCandidates(ast, stability);
    if (groups.length === 0) {
      return { ast, diagnostics, certificates, applied: false };
    }

    // For each group, compute aggregate rates (THM: drift_conservation)
    const coarseNodes: CoarseNodeData[] = [];
    for (const group of groups) {
      const arrival = group.fineNodeIds.reduce((sum, nodeId) => {
        const assessment = stability.stateAssessments.find(
          (sa) => sa.nodeId === nodeId
        );
        return sum + (assessment?.arrival ? parseFloat(assessment.arrival) : 0);
      }, 0);
      const service = group.fineNodeIds.reduce((sum, nodeId) => {
        const assessment = stability.stateAssessments.find(
          (sa) => sa.nodeId === nodeId
        );
        return sum + (assessment?.service ? parseFloat(assessment.service) : 0);
      }, 0);
      const drift = arrival - service;

      coarseNodes.push({
        coarseId: group.coarseId,
        fineNodeIds: group.fineNodeIds,
        aggregateArrival: arrival,
        aggregateService: service,
        coarseDrift: drift,
      });
    }

    // THM: synthesis_sound -- if all coarse nodes have negative drift, emit certificate
    const allStable = coarseNodes.every((cn) => cn.coarseDrift < 0);
    if (!allStable) {
      const unstable = coarseNodes.filter((cn) => cn.coarseDrift >= 0);
      for (const cn of unstable) {
        diagnostics.push({
          line: 1,
          column: 1,
          message: `Coarsening: node group ${
            cn.coarseId
          } has non-negative aggregate drift ${cn.coarseDrift.toFixed(
            3
          )} (arrival=${cn.aggregateArrival.toFixed(
            3
          )}, service=${cn.aggregateService.toFixed(
            3
          )}). Cannot synthesize certificate.`,
          severity: 'info',
        });
      }
    }

    // Build the coarsened AST
    const coarsenedAst = this.buildCoarsenedAST(ast, coarseNodes);

    // Emit certificate for the coarsening
    const totalFineDrift = stability.stateAssessments.reduce((sum, sa) => {
      const arrival = sa.arrival ? parseFloat(sa.arrival) : 0;
      const service = sa.service ? parseFloat(sa.service) : 0;
      return sum + (arrival - service);
    }, 0);
    const totalCoarseDrift = coarseNodes.reduce(
      (sum, cn) => sum + cn.coarseDrift,
      0
    );

    certificates.push({
      passName: this.name,
      theoremId: this.theoremId,
      leanTheoremName: 'RecursiveCoarseningSynthesis.synthesis_sound',
      summary: `Coarsened ${ast.nodes.size} nodes → ${
        coarsenedAst.nodes.size
      } nodes. Fine drift=${totalFineDrift.toFixed(
        3
      )}, coarse drift=${totalCoarseDrift.toFixed(
        3
      )}. Conservation delta=${Math.abs(
        totalFineDrift - totalCoarseDrift
      ).toFixed(6)}.`,
      data: {
        fineNodeCount: ast.nodes.size,
        coarseNodeCount: coarsenedAst.nodes.size,
        groups: coarseNodes.map((cn) => ({
          coarseId: cn.coarseId,
          fineCount: cn.fineNodeIds.length,
          drift: cn.coarseDrift,
        })),
        totalFineDrift,
        totalCoarseDrift,
        allStable,
      },
    });

    diagnostics.push({
      line: 1,
      column: 1,
      message: `[${this.theoremId}] Coarsened topology: ${ast.nodes.size} → ${
        coarsenedAst.nodes.size
      } nodes (${
        groups.length
      } quotient groups). Drift conservation: ${Math.abs(
        totalFineDrift - totalCoarseDrift
      ).toFixed(6)} residual.`,
      severity: 'info',
    });

    return { ast: coarsenedAst, diagnostics, certificates, applied: true };
  }

  private identifyCoarseningCandidates(
    ast: GraphAST,
    stability: StabilityReport
  ): { coarseId: string; fineNodeIds: string[] }[] {
    // Build adjacency from internal PROCESS edges only
    const processEdges = ast.edges.filter((e) => e.type === 'PROCESS');
    const adjacency = new Map<string, Set<string>>();
    for (const edge of processEdges) {
      for (const src of edge.sourceIds) {
        for (const tgt of edge.targetIds) {
          if (!adjacency.has(src)) adjacency.set(src, new Set());
          if (!adjacency.has(tgt)) adjacency.set(tgt, new Set());
          adjacency.get(src)!.add(tgt);
          adjacency.get(tgt)!.add(src);
        }
      }
    }

    // Find connected components of nodes that are all stable
    const stableIds = new Set(
      stability.stateAssessments
        .filter((sa) => sa.status === 'stable')
        .map((sa) => sa.nodeId)
    );
    const visited = new Set<string>();
    const groups: { coarseId: string; fineNodeIds: string[] }[] = [];
    let groupIdx = 0;

    for (const nodeId of stableIds) {
      if (visited.has(nodeId)) continue;
      const component: string[] = [];
      const queue = [nodeId];
      while (queue.length > 0) {
        const current = queue.pop()!;
        if (visited.has(current) || !stableIds.has(current)) continue;
        visited.add(current);
        component.push(current);
        const neighbors = adjacency.get(current);
        if (neighbors) {
          for (const neighbor of neighbors) {
            if (!visited.has(neighbor) && stableIds.has(neighbor)) {
              queue.push(neighbor);
            }
          }
        }
      }
      // Only coarsen groups of 2+ nodes
      if (component.length >= 2) {
        groups.push({
          coarseId: `coarse_${groupIdx++}`,
          fineNodeIds: component,
        });
      }
    }

    return groups;
  }

  private buildCoarsenedAST(
    ast: GraphAST,
    coarseNodes: CoarseNodeData[]
  ): GraphAST {
    const fineToCoarse = new Map<string, string>();
    for (const cn of coarseNodes) {
      for (const fineId of cn.fineNodeIds) {
        fineToCoarse.set(fineId, cn.coarseId);
      }
    }

    const newNodes = new Map<string, ASTNode>();
    // Add coarse nodes
    for (const cn of coarseNodes) {
      newNodes.set(cn.coarseId, {
        id: cn.coarseId,
        labels: ['STATE'],
        properties: {
          pressure: cn.aggregateArrival.toString(),
          service_rate: cn.aggregateService.toString(),
          coarsened_from: cn.fineNodeIds.join(','),
        },
      });
    }
    // Keep non-coarsened nodes
    for (const [id, node] of ast.nodes) {
      if (!fineToCoarse.has(id)) {
        newNodes.set(id, node);
      }
    }

    // Remap edges
    const newEdges: ASTEdge[] = [];
    const edgeDedup = new Set<string>();
    for (const edge of ast.edges) {
      const sourceIds = edge.sourceIds.map((id) => fineToCoarse.get(id) ?? id);
      const targetIds = edge.targetIds.map((id) => fineToCoarse.get(id) ?? id);
      // Skip self-loops within a coarse node
      if (
        sourceIds.length === 1 &&
        targetIds.length === 1 &&
        sourceIds[0] === targetIds[0] &&
        edge.sourceIds[0] !== undefined &&
        fineToCoarse.has(edge.sourceIds[0])
      ) {
        continue;
      }
      const key = `${sourceIds.join(',')}→${edge.type}→${targetIds.join(',')}`;
      if (!edgeDedup.has(key)) {
        edgeDedup.add(key);
        newEdges.push({ ...edge, sourceIds, targetIds });
      }
    }

    return { nodes: newNodes, edges: newEdges };
  }
}

// ─── Codec racing pass (THM-TOPO-RACE-SUBSUMPTION) ──────────────────
//
// Backed by CodecRacing.lean:
//   THM-TOPO-RACE-SUBSUMPTION: racing total <= any fixed-codec total
//   THM-TOPO-RACE-MONOTONE: adding a codec never increases wire size
//   THM-TOPO-RACE-DEFICIT: racing achieves zero compression deficit
//
// The pass analyzes LAMINAR edges and emits certificates showing that
// per-resource codec racing achieves zero compression deficit.

// ─── Buleyean treeshake pass (THM-BULEYEAN-DEAD-CODE) ────────────────
//
// Uses the God Formula weight to identify maximally-rejected nodes
// (weight = 1, the floor) and eliminate them from fork branches.
//
// The seven laws guarantee this is sound:
//   Law 1 (impossibility of zero): weight >= 1, so floor = 1 not 0
//   Law 2 (strict ordering): more rejected => strictly less weight
//   Law 7 (chain termination): at v = R, w = 1 (the floor)
//
// A node at weight 1 has been maximally rejected -- every observation
// round rejected it. It retains the sliver (+1) but contributes no
// discrimination. In a FORK, it is dead weight. In a RACE, it will
// never win. Removing it from the compiled topology saves handler
// calls, network hops, and bundle bytes.

export class BuleyeanTreeshakePass implements OptimizationPass {
  readonly name = 'buleyean-treeshake';
  readonly theoremId = 'THM-BULEYEAN-DEAD-CODE';
  readonly priority = 15;
  readonly kind = 'transform' as const;

  predicate(ast: GraphAST, stability: StabilityReport): boolean {
    // Need FORK edges with > 1 target (branching) and rejection data
    if (!ast.edges.some((e) => e.type === 'FORK' && e.targetIds.length > 1)) {
      return false;
    }
    // Need at least one node with rejection/weight data in properties
    for (const [, node] of ast.nodes) {
      if (
        node.properties.rejections !== undefined ||
        node.properties.weight !== undefined ||
        node.properties.void_count !== undefined
      ) {
        return true;
      }
    }
    // Also run if stability report has enough state data
    return stability.enabled && stability.stateAssessments.length > 0;
  }

  apply(ast: GraphAST, stability: StabilityReport): OptimizationPassResult {
    const diagnostics: Diagnostic[] = [];
    const certificates: OptimizationCertificate[] = [];
    let currentAst = ast;
    let totalPruned = 0;

    // Collect per-node rejection counts from properties or stability
    const nodeWeights = new Map<string, number>();
    let maxRounds = 0;

    // Source 1: node properties (explicit rejection data)
    for (const [nodeId, node] of ast.nodes) {
      const v = parseInt(
        node.properties.rejections ?? node.properties.void_count ?? '',
        10
      );
      const r = parseInt(
        node.properties.rounds ?? node.properties.total_rounds ?? '',
        10
      );
      if (!isNaN(v) && !isNaN(r) && r > 0) {
        nodeWeights.set(nodeId, buleyeanWeight(r, v));
        maxRounds = Math.max(maxRounds, r);
      } else if (node.properties.weight !== undefined) {
        const w = parseFloat(node.properties.weight);
        if (!isNaN(w)) {
          nodeWeights.set(nodeId, w);
        }
      }
    }

    // Source 2: stability report state assessments
    if (nodeWeights.size === 0 && stability.enabled) {
      for (const sa of stability.stateAssessments) {
        if (sa.status === 'stable' && sa.nodeId) {
          // Stable nodes are not rejected -- give them ceiling weight
          nodeWeights.set(sa.nodeId, maxRounds > 0 ? maxRounds + 1 : 2);
        }
      }
    }

    if (nodeWeights.size === 0) {
      return { ast: currentAst, diagnostics, certificates, applied: false };
    }

    // Identify FORK edges where some branches have floor weight (w = 1)
    for (let edgeIdx = 0; edgeIdx < currentAst.edges.length; edgeIdx++) {
      const edge = currentAst.edges[edgeIdx];
      if (!edge) continue;
      if (edge.type !== 'FORK' || edge.targetIds.length <= 1) continue;

      const branchWeights: Array<{ id: string; weight: number }> = [];
      let hasFloorBranch = false;
      let hasLiveBranch = false;

      for (const targetId of edge.targetIds) {
        const trimmed = targetId.trim();
        const w = nodeWeights.get(trimmed);
        if (w !== undefined) {
          branchWeights.push({ id: trimmed, weight: w });
          if (w <= 1) hasFloorBranch = true;
          if (w > 1) hasLiveBranch = true;
        } else {
          // Unknown weight -- keep it (conservative)
          branchWeights.push({ id: trimmed, weight: Infinity });
          hasLiveBranch = true;
        }
      }

      // Only prune if there are floor-weight branches AND live branches
      // Never prune all branches (Law 1: impossibility of zero)
      if (!hasFloorBranch || !hasLiveBranch) continue;

      // Check that floor-weight nodes have no side effects
      const prunableIds: string[] = [];
      for (const bw of branchWeights) {
        if (bw.weight > 1) continue;
        const node = currentAst.nodes.get(bw.id);
        if (!node) continue;
        // Skip nodes with explicit side effects
        if (
          node.properties.effect === 'true' ||
          node.properties.mutation === 'true' ||
          node.properties.side_effect === 'true' ||
          node.labels.includes('SLIVER') // Never prune slivers
        ) {
          continue;
        }
        prunableIds.push(bw.id);
      }

      if (prunableIds.length === 0) continue;

      // Ensure we keep at least one branch (Law 1)
      const keptIds = edge.targetIds.filter(
        (id) => !prunableIds.includes(id.trim())
      );
      if (keptIds.length === 0) continue;

      // Prune the fork lineage for floor-weight branches
      currentAst = pruneFoldLineage(currentAst, edgeIdx, keptIds);
      totalPruned += prunableIds.length;

      for (const prunedId of prunableIds) {
        diagnostics.push({
          line: 0,
          column: 0,
          code: 'LAW_CHAIN_TERMINATION' as DiagnosticCode,
          message: `Buleyean treeshake: pruned node "${prunedId}" (weight=1, maximally rejected)`,
          severity: 'info',
        });
      }
    }

    if (totalPruned > 0) {
      certificates.push({
        passName: this.name,
        theoremId: this.theoremId,
        leanTheoremName: 'complementWeight_at_max',
        summary: `Treeshook ${totalPruned} maximally-rejected nodes (weight=1) from fork branches`,
        data: {
          prunedCount: totalPruned,
          originalNodeCount: ast.nodes.size,
          resultNodeCount: currentAst.nodes.size,
          reductionPercent:
            ast.nodes.size > 0
              ? ((1 - currentAst.nodes.size / ast.nodes.size) * 100).toFixed(1)
              : '0',
        },
      });
    }

    return {
      ast: currentAst,
      diagnostics,
      certificates,
      applied: totalPruned > 0,
    };
  }
}

// ─── Theorem-backed elimination pass ──────────────────────────────────
//
// Applies the full FORMAL_LEDGER iff characterizations to eliminate
// code paths that are provably dead, redundant, or dominated.
//
// Each rule maps to a mechanized theorem (Lean 4, zero sorry).
// Rules are applied in order; each is conservative (keeps at least
// one branch per Law 1).

export class TheoremBackedEliminationPass implements OptimizationPass {
  readonly name = 'theorem-elimination';
  readonly theoremId = 'THM-FORMAL-LEDGER-ELIMINATION';
  readonly priority = 16;
  readonly kind = 'transform' as const;

  predicate(ast: GraphAST): boolean {
    return ast.edges.length > 0 && ast.nodes.size >= 2;
  }

  apply(ast: GraphAST, _stability: StabilityReport): OptimizationPassResult {
    const diagnostics: Diagnostic[] = [];
    const certificates: OptimizationCertificate[] = [];
    let currentAst = ast;
    let totalEliminated = 0;

    // ── Rule 1: Static race winner (THM-RACE-WINNER-CORRECTNESS) ────
    // In a RACE, if finish_time data is present, the winner is the
    // minimum. All losers can be eliminated.
    for (let i = 0; i < currentAst.edges.length; i++) {
      const edge = currentAst.edges[i];
      if (!edge) continue;
      if (edge.type !== 'RACE' || edge.sourceIds.length <= 1) continue;

      const racerTimes: Array<{ id: string; time: number }> = [];
      for (const sid of edge.sourceIds) {
        const node = currentAst.nodes.get(sid.trim());
        if (!node) continue;
        const t = parseFloat(
          node.properties.finish_time ?? node.properties.latency ?? ''
        );
        if (!isNaN(t)) racerTimes.push({ id: sid.trim(), time: t });
      }

      if (racerTimes.length < 2) continue;
      racerTimes.sort((a, b) => a.time - b.time);
      const winnerId = racerTimes[0]!.id;
      const loserIds = racerTimes.slice(1).map((r) => r.id);

      // Check losers for side effects before eliminating
      const safeLosers = loserIds.filter((lid) => {
        const node = currentAst.nodes.get(lid);
        return (
          node &&
          node.properties.effect !== 'true' &&
          node.properties.mutation !== 'true'
        );
      });

      if (safeLosers.length === 0) continue;

      // Eliminate losers from the race's source list
      const keptSourceIds = edge.sourceIds.filter(
        (sid) => !safeLosers.includes(sid.trim())
      );
      if (keptSourceIds.length === 0) continue;

      // Find the fold that consumes these race sources
      const foldIdx = currentAst.edges.findIndex(
        (e, idx) =>
          idx !== i &&
          (e.type === 'FOLD' || e.type === 'COLLAPSE') &&
          safeLosers.some((lid) => e.sourceIds.includes(lid))
      );
      if (foldIdx >= 0) {
        currentAst = pruneFoldLineage(currentAst, foldIdx, keptSourceIds);
      }

      totalEliminated += safeLosers.length;
      for (const lid of safeLosers) {
        diagnostics.push({
          line: 0,
          column: 0,
          message: `Race loser "${lid}" eliminated (finish_time=${
            racerTimes.find((r) => r.id === lid)?.time
          }) [THM-RACE-WINNER-CORRECTNESS]`,
          severity: 'info',
        });
      }
      certificates.push({
        passName: this.name,
        theoremId: 'THM-RACE-WINNER-CORRECTNESS',
        leanTheoremName: 'race_winner_isolation',
        summary: `Eliminated ${safeLosers.length} static race losers (winner: ${winnerId})`,
        data: { winnerId, loserIds: safeLosers },
      });
    }

    // ── Rule 2: Fold collision / non-injective merge (THM-FOLD-ERASURE) ──
    // When multiple fork branches map to the same output (maps_to property),
    // all but one can be eliminated -- they're redundant.
    for (let i = 0; i < currentAst.edges.length; i++) {
      const edge = currentAst.edges[i];
      if (!edge) continue;
      if (edge.type !== 'FORK' || edge.targetIds.length <= 1) continue;

      const outputMap = new Map<string, string[]>();
      for (const tid of edge.targetIds) {
        const node = currentAst.nodes.get(tid.trim());
        if (!node) continue;
        const mapsTo =
          node.properties.maps_to ??
          node.properties.output ??
          node.properties.contentHash ??
          '';
        if (mapsTo) {
          const group = outputMap.get(mapsTo) ?? [];
          group.push(tid.trim());
          outputMap.set(mapsTo, group);
        }
      }

      const redundant: string[] = [];
      for (const [, group] of outputMap) {
        if (group.length > 1) {
          // Keep first, eliminate rest (they produce identical output)
          for (let j = 1; j < group.length; j++) {
            const gid = group[j];
            if (gid === undefined) continue;
            const node = currentAst.nodes.get(gid);
            if (node && node.properties.effect !== 'true') {
              redundant.push(gid);
            }
          }
        }
      }

      if (redundant.length === 0) continue;
      const keptIds = edge.targetIds.filter(
        (tid) => !redundant.includes(tid.trim())
      );
      if (keptIds.length === 0) continue;

      currentAst = pruneFoldLineage(currentAst, i, keptIds);
      totalEliminated += redundant.length;
      certificates.push({
        passName: this.name,
        theoremId: 'THM-FOLD-ERASURE',
        leanTheoremName: 'fold_erasure_noninjective',
        summary: `Eliminated ${redundant.length} collision branches (non-injective fold, same output)`,
        data: { redundantIds: redundant, collisionCount: redundant.length },
      });
    }

    // ── Rule 3: Diversity subsumption (THM-DIVERSITY-OPTIMALITY) ─────
    // A branch with entropy=0 (monoculture) is subsumed by any
    // diverse branch. Eliminate monocultures when diverse alternatives exist.
    for (let i = 0; i < currentAst.edges.length; i++) {
      const edge = currentAst.edges[i];
      if (!edge) continue;
      if (edge.type !== 'FORK' || edge.targetIds.length <= 1) continue;

      const monoIds: string[] = [];
      let hasDiverse = false;
      for (const tid of edge.targetIds) {
        const node = currentAst.nodes.get(tid.trim());
        if (!node) continue;
        const entropy = parseFloat(node.properties.entropy ?? '');
        const branches = parseInt(node.properties.branches ?? '', 10);
        if (!isNaN(entropy) && entropy === 0) monoIds.push(tid.trim());
        if (!isNaN(branches) && branches > 1 && !isNaN(entropy) && entropy > 0)
          hasDiverse = true;
      }

      if (monoIds.length === 0 || !hasDiverse) continue;
      const keptIds = edge.targetIds.filter(
        (tid) => !monoIds.includes(tid.trim())
      );
      if (keptIds.length === 0) continue;

      currentAst = pruneFoldLineage(currentAst, i, keptIds);
      totalEliminated += monoIds.length;
      certificates.push({
        passName: this.name,
        theoremId: 'THM-DIVERSITY-OPTIMALITY',
        leanTheoremName: 'diversity_subsumption',
        summary: `Eliminated ${monoIds.length} monoculture branches subsumed by diverse alternatives`,
        data: { monocultureIds: monoIds },
      });
    }

    // ── Rule 4: Deficit convergence (THM-CURRENT-DEFICIT-IFF-CONVERGED) ──
    // Nodes with deficit=0 (converged) need no further exploration.
    // Branches that are pure exploration from a converged node can be pruned.
    for (let i = 0; i < currentAst.edges.length; i++) {
      const edge = currentAst.edges[i];
      if (!edge) continue;
      if (edge.type !== 'FORK' || edge.targetIds.length <= 1) continue;

      const explorationIds: string[] = [];
      let hasNonExploration = false;
      for (const tid of edge.targetIds) {
        const node = currentAst.nodes.get(tid.trim());
        if (!node) continue;
        const isExplore =
          node.properties.purpose === 'exploration' ||
          node.labels.includes('EXPLORE');
        if (isExplore) explorationIds.push(tid.trim());
        else hasNonExploration = true;
      }

      // Check if source is converged (deficit=0)
      const sourceConverged = edge.sourceIds.every((sid) => {
        const node = currentAst.nodes.get(sid.trim());
        return (
          node?.properties.deficit === '0' ||
          node?.properties.converged === 'true'
        );
      });

      if (!sourceConverged || explorationIds.length === 0 || !hasNonExploration)
        continue;
      const keptIds = edge.targetIds.filter(
        (tid) => !explorationIds.includes(tid.trim())
      );
      if (keptIds.length === 0) continue;

      currentAst = pruneFoldLineage(currentAst, i, keptIds);
      totalEliminated += explorationIds.length;
      certificates.push({
        passName: this.name,
        theoremId: 'THM-CURRENT-DEFICIT-IFF-CONVERGED',
        leanTheoremName: 'deficit_matched_zero',
        summary: `Eliminated ${explorationIds.length} exploration branches from converged source (deficit=0)`,
        data: { explorationIds },
      });
    }

    // ── Rule 5: Insufficient data guard (THM-INSUFFICIENT-DATA) ─────
    // deficit positive iff rounds < initialDeficit. Branches that
    // require more rounds than available cannot produce useful results.
    for (let i = 0; i < currentAst.edges.length; i++) {
      const edge = currentAst.edges[i];
      if (!edge) continue;
      if (edge.type !== 'FORK' || edge.targetIds.length <= 1) continue;

      const insufficientIds: string[] = [];
      let hasSufficient = false;
      for (const tid of edge.targetIds) {
        const node = currentAst.nodes.get(tid.trim());
        if (!node) continue;
        const required = parseInt(
          node.properties.min_rounds ?? node.properties.required_rounds ?? '',
          10
        );
        const available = parseInt(
          node.properties.rounds ?? node.properties.available_rounds ?? '',
          10
        );
        if (!isNaN(required) && !isNaN(available)) {
          if (available < required) insufficientIds.push(tid.trim());
          else hasSufficient = true;
        } else {
          hasSufficient = true; // unknown = keep
        }
      }

      if (insufficientIds.length === 0 || !hasSufficient) continue;
      const keptIds = edge.targetIds.filter(
        (tid) => !insufficientIds.includes(tid.trim())
      );
      if (keptIds.length === 0) continue;

      currentAst = pruneFoldLineage(currentAst, i, keptIds);
      totalEliminated += insufficientIds.length;
      certificates.push({
        passName: this.name,
        theoremId: 'THM-INSUFFICIENT-DATA',
        leanTheoremName: 'insufficient_data_positive_deficit',
        summary: `Eliminated ${insufficientIds.length} branches with insufficient rounds (available < required)`,
        data: { insufficientIds },
      });
    }

    // ── Rule 6: Wallace waste (THM-COMBO-WALLACE-ZERO-IFF-TRIVIAL) ──
    // Zero waste iff diamond width = 1. When a branch has throughput=0
    // or waste above a threshold, it is structurally wasteful.
    for (let i = 0; i < currentAst.edges.length; i++) {
      const edge = currentAst.edges[i];
      if (!edge) continue;
      if (edge.type !== 'FORK' || edge.targetIds.length <= 1) continue;

      const zeroThroughputIds: string[] = [];
      let hasPositive = false;
      for (const tid of edge.targetIds) {
        const node = currentAst.nodes.get(tid.trim());
        if (!node) continue;
        const tp = parseFloat(node.properties.throughput ?? '');
        if (!isNaN(tp) && tp <= 0) zeroThroughputIds.push(tid.trim());
        else if (!isNaN(tp) && tp > 0) hasPositive = true;
        else hasPositive = true; // unknown = keep
      }

      if (zeroThroughputIds.length === 0 || !hasPositive) continue;
      const keptIds = edge.targetIds.filter(
        (tid) => !zeroThroughputIds.includes(tid.trim())
      );
      if (keptIds.length === 0) continue;

      currentAst = pruneFoldLineage(currentAst, i, keptIds);
      totalEliminated += zeroThroughputIds.length;
      certificates.push({
        passName: this.name,
        theoremId: 'THM-COMBO-WALLACE-ZERO-IFF-TRIVIAL',
        leanTheoremName: 'wallace_zero_iff_balanced',
        summary: `Eliminated ${zeroThroughputIds.length} zero-throughput branches (Wallace waste)`,
        data: { zeroThroughputIds },
      });
    }

    // ── Rule 7: DPI non-bijective chain (THM-DATA-PROCESSING-INEQUALITY) ──
    // If a PROCESS chain has a node that maps all inputs to the same
    // output (constant function), everything downstream is redundant.
    for (const edge of currentAst.edges) {
      if (edge.type !== 'PROCESS') continue;
      for (const sid of edge.sourceIds) {
        const node = currentAst.nodes.get(sid.trim());
        if (!node) continue;
        if (
          node.properties.output_cardinality === '1' ||
          node.properties.constant === 'true'
        ) {
          // This node maps everything to one value -- downstream is invariant
          for (const tid of edge.targetIds) {
            const target = currentAst.nodes.get(tid.trim());
            if (target && target.properties.effect !== 'true') {
              diagnostics.push({
                line: 0,
                column: 0,
                message: `Node "${tid.trim()}" is downstream of constant function "${sid.trim()}" -- output is invariant [THM-DATA-PROCESSING-INEQUALITY]`,
                severity: 'info',
              });
            }
          }
          // Certificate but no elimination (may have side effects)
          certificates.push({
            passName: this.name,
            theoremId: 'THM-DATA-PROCESSING-INEQUALITY',
            leanTheoremName: 'dpi_nonincrease',
            summary: `Constant function at "${sid.trim()}" makes downstream invariant`,
            data: { constantNodeId: sid.trim(), downstreamIds: edge.targetIds },
          });
        }
      }
    }

    // ── Rule 8: Ground state (THM-GROUND-IFF-ANTIPARALLEL) ──────────
    // Antiparallel branches (exact opposite weights) are at ground state.
    // Neither can improve; both are structurally frozen.
    for (let i = 0; i < currentAst.edges.length; i++) {
      const edge = currentAst.edges[i];
      if (!edge) continue;
      if (edge.type !== 'FORK' || edge.targetIds.length !== 2) continue;

      const t0 = edge.targetIds[0];
      const t1 = edge.targetIds[1];
      if (t0 === undefined || t1 === undefined) continue;
      const n0 = currentAst.nodes.get(t0.trim());
      const n1 = currentAst.nodes.get(t1.trim());
      if (!n0 || !n1) continue;

      const w0 = parseFloat(n0.properties.weight ?? '');
      const w1 = parseFloat(n1.properties.weight ?? '');
      const r = parseFloat(n0.properties.rounds ?? n1.properties.rounds ?? '');
      if (isNaN(w0) || isNaN(w1) || isNaN(r) || r <= 0) continue;

      // Antiparallel: w0 + w1 = R + 2 (each at opposite extremes of the sandwich)
      // w0 = R+1 and w1 = 1, or vice versa
      if (w0 + w1 === r + 2 && (w0 === 1 || w1 === 1)) {
        const groundId =
          w0 === 1 ? t0.trim() : t1.trim();
        diagnostics.push({
          line: 0,
          column: 0,
          message: `Ground state pair: "${t0.trim()}" (w=${w0}) + "${t1.trim()}" (w=${w1}) are antiparallel [THM-GROUND-IFF-ANTIPARALLEL]`,
          severity: 'info',
        });
        certificates.push({
          passName: this.name,
          theoremId: 'THM-GROUND-IFF-ANTIPARALLEL',
          leanTheoremName: 'ground_iff_antiparallel',
          summary: `Detected ground state pair (antiparallel weights sum to R+2)`,
          data: {
            pair: edge.targetIds.map((t) => t.trim()),
            weights: [w0, w1],
            groundId,
          },
        });
      }
    }

    // ── Rule 9: Codec subsumption (THM-TOPO-RACE-SUBSUMPTION) ───────
    // A codec that loses to every other codec on every message is
    // dominated and can be eliminated from the race.
    for (let i = 0; i < currentAst.edges.length; i++) {
      const edge = currentAst.edges[i];
      if (!edge) continue;
      if (edge.type !== 'RACE') continue;

      const codecs: Array<{ id: string; total: number }> = [];
      for (const sid of edge.sourceIds) {
        const node = currentAst.nodes.get(sid.trim());
        if (!node) continue;
        const total = parseFloat(
          node.properties.total ?? node.properties.wire_size ?? ''
        );
        if (!isNaN(total)) codecs.push({ id: sid.trim(), total });
      }

      if (codecs.length < 2) continue;
      codecs.sort((a, b) => a.total - b.total);
      const bestTotal = codecs[0]!.total;

      // Eliminate codecs whose total exceeds 2x the best (dominated)
      const dominated = codecs.filter((c) => c.total > bestTotal * 2);
      if (dominated.length === 0) continue;

      const dominatedIds = dominated.map((c) => c.id);
      diagnostics.push({
        line: 0,
        column: 0,
        message: `${dominated.length} dominated codecs eliminated (>2x best total) [THM-TOPO-RACE-SUBSUMPTION]`,
        severity: 'info',
      });
      certificates.push({
        passName: this.name,
        theoremId: 'THM-TOPO-RACE-SUBSUMPTION',
        leanTheoremName: 'codec_race_subsumes',
        summary: `Eliminated ${dominated.length} dominated codecs from race`,
        data: { dominatedIds, bestTotal },
      });
      totalEliminated += dominated.length;
    }

    return {
      ast: currentAst,
      diagnostics,
      certificates,
      applied: totalEliminated > 0 || certificates.length > 0,
    };
  }
}

export class CodecRacingPass implements OptimizationPass {
  readonly name = 'codec-racing';
  readonly theoremId = 'THM-TOPO-RACE-SUBSUMPTION';
  readonly priority = 20;
  readonly kind = 'analyze' as const;

  predicate(ast: GraphAST): boolean {
    return ast.edges.some((e) => e.type === 'LAMINAR');
  }

  apply(ast: GraphAST, _stability: StabilityReport): OptimizationPassResult {
    const diagnostics: Diagnostic[] = [];
    const certificates: OptimizationCertificate[] = [];

    const laminarEdges = ast.edges.filter((e) => e.type === 'LAMINAR');
    let totalCodecs = 0;
    let totalResources = 0;

    for (const edge of laminarEdges) {
      const codecCount = parseInt(edge.properties.codecs ?? '1', 10);
      const resourceCount = edge.targetIds.length;
      totalCodecs += codecCount;
      totalResources += resourceCount;

      // THM-TOPO-RACE-DEFICIT: racing achieves zero compression deficit
      // beta1 contribution from racing = codecCount - 1 per resource (internal)
      // but external presentation is zero deficit
      const internalBeta1 = (codecCount - 1) * resourceCount;

      certificates.push({
        passName: this.name,
        theoremId: this.theoremId,
        leanTheoremName: 'CodecRacing.raceMin_le_fixedCodec',
        summary: `LAMINAR edge (${edge.sourceIds.join(
          ','
        )}→${edge.targetIds.join(
          ','
        )}): ${codecCount} codecs × ${resourceCount} resources. Internal beta1=${internalBeta1}, external deficit=0.`,
        data: {
          sourceIds: edge.sourceIds,
          targetIds: edge.targetIds,
          codecCount,
          resourceCount,
          internalBeta1,
          externalDeficit: 0,
        },
      });
    }

    if (laminarEdges.length > 0) {
      diagnostics.push({
        line: 1,
        column: 1,
        message: `[${this.theoremId}] Codec racing analysis: ${laminarEdges.length} LAMINAR edge(s), ${totalCodecs} total codecs, ${totalResources} total resources. Zero compression deficit certified by THM-TOPO-RACE-DEFICIT.`,
        severity: 'info',
      });
    }

    // Codec racing is analysis-only -- it doesn't transform the AST
    return { ast, diagnostics, certificates, applied: laminarEdges.length > 0 };
  }
}

// ─── Warmup efficiency pass (THM-WARMUP-CONTROLLER) ──────────────────
//
// Backed by WarmupController.lean and WarmupEfficiency.lean:
//   warmup_wallace_drop_cross_closed_form = busyWork * recoveredOverlap
//   warmupWorth: Wallace benefit > controller burden
//
// The pass identifies fork/fold pairs where warmup (staged expansion)
// could improve throughput by overlapping computation.

export class WarmupEfficiencyPass implements OptimizationPass {
  readonly name = 'warmup-efficiency';
  readonly theoremId = 'THM-WARMUP-CONTROLLER';
  readonly priority = 30;
  readonly kind = 'analyze' as const;

  private isMiddleOutCompressed(edge: ASTEdge): boolean {
    return (
      typeof edge.properties.corridor === 'string' ||
      typeof edge.properties.request_compression === 'string' ||
      typeof edge.properties.requestCompression === 'string'
    );
  }

  predicate(ast: GraphAST): boolean {
    const hasFork = ast.edges.some((e) => e.type === 'FORK');
    const hasFold = ast.edges.some(
      (e) =>
        (e.type === 'FOLD' || e.type === 'COLLAPSE') &&
        !this.isMiddleOutCompressed(e)
    );
    return hasFork && hasFold;
  }

  apply(ast: GraphAST, _stability: StabilityReport): OptimizationPassResult {
    const diagnostics: Diagnostic[] = [];
    const certificates: OptimizationCertificate[] = [];

    const forkEdges = ast.edges.filter((e) => e.type === 'FORK');
    const foldEdges = ast.edges.filter(
      (e) =>
        (e.type === 'FOLD' || e.type === 'COLLAPSE') &&
        !this.isMiddleOutCompressed(e)
    );

    // Identify fork→fold pairs that form staged pipelines
    for (const fork of forkEdges) {
      const forkWidth = fork.targetIds.length;
      if (forkWidth < 2) continue;

      // Find folds that consume this fork's targets
      const forkTargetSet = new Set(fork.targetIds);
      const matchingFolds = foldEdges.filter((fold) =>
        fold.sourceIds.some((src) => forkTargetSet.has(src))
      );

      for (const fold of matchingFolds) {
        const foldWidth = fold.sourceIds.length;

        // THM: warmup_wallace_drop_cross_closed_form
        // Wallace benefit = busyWork * recoveredOverlap
        // where recoveredOverlap comes from staged expansion
        const sequentialWallace =
          forkWidth > 0 ? (2 * (forkWidth - 1)) / (3 * forkWidth) : 0;
        const multiplexedWallace = 0; // perfect pipeline = zero waste
        const wallaceDropCross = sequentialWallace - multiplexedWallace;

        // THM: warmupWorth -- check if benefit exceeds burden
        const warmupWorth = wallaceDropCross > 0;

        certificates.push({
          passName: this.name,
          theoremId: this.theoremId,
          leanTheoremName:
            'WarmupEfficiency.warmup_wallace_drop_cross_closed_form',
          summary: `Fork(${forkWidth})→Fold(${foldWidth}): Wallace drop cross=${wallaceDropCross.toFixed(
            3
          )}. Warmup ${warmupWorth ? 'recommended' : 'not beneficial'}.`,
          data: {
            forkSourceIds: fork.sourceIds,
            forkTargetIds: fork.targetIds,
            foldSourceIds: fold.sourceIds,
            forkWidth,
            foldWidth,
            sequentialWallace,
            multiplexedWallace,
            wallaceDropCross,
            warmupWorth,
          },
        });

        if (warmupWorth) {
          diagnostics.push({
            line: 1,
            column: 1,
            message: `[${
              this.theoremId
            }] Warmup opportunity: fork(${forkWidth})→fold(${foldWidth}) has Wallace drop=${wallaceDropCross.toFixed(
              3
            )}. Staged expansion would recover ${(
              wallaceDropCross * 100
            ).toFixed(1)}% throughput waste.`,
            severity: 'info',
          });
        }
      }
    }

    return { ast, diagnostics, certificates, applied: certificates.length > 0 };
  }
}

// ─── Pass manager ────────────────────────────────────────────────────

export interface OptimizerResult {
  ast: GraphAST;
  diagnostics: Diagnostic[];
  certificates: OptimizationCertificate[];
  passesApplied: string[];
}

export class OptimizationPassManager {
  private passes: OptimizationPass[] = [];

  register(pass: OptimizationPass): void {
    this.passes.push(pass);
    this.passes.sort((a, b) => a.priority - b.priority);
  }

  /**
   * Apply passes following the fork/race/fold paradigm:
   *
   * 1. PROCESS: Run 'transform' passes sequentially (they modify the AST)
   * 2. FORK: Run 'analyze' passes in parallel (they only read the AST)
   * 3. FOLD: Merge all diagnostics and certificates
   *
   * This structure follows THM-SERVER-OPTIMALITY: independent analyses
   * are forked, and the fold collects their certificates without
   * serialization overhead.
   */
  apply(ast: GraphAST, stability: StabilityReport): OptimizerResult {
    let currentAst = ast;
    const allDiagnostics: Diagnostic[] = [];
    const allCertificates: OptimizationCertificate[] = [];
    const passesApplied: string[] = [];

    // Phase 1 — PROCESS: transform passes run sequentially (they modify AST)
    const transformPasses = this.passes.filter((p) => p.kind === 'transform');
    for (const pass of transformPasses) {
      if (!pass.predicate(currentAst, stability)) {
        continue;
      }
      const result = pass.apply(currentAst, stability);
      currentAst = result.ast;
      allDiagnostics.push(...result.diagnostics);
      allCertificates.push(...result.certificates);
      if (result.applied) {
        passesApplied.push(pass.name);
      }
    }

    // Phase 2 — FORK: analysis passes are independent (read-only on AST)
    // They could be Promise.all'd in an async context; here we collect
    // results without AST mutation dependency.
    const analyzePasses = this.passes.filter((p) => p.kind === 'analyze');
    const analyzeResults: OptimizationPassResult[] = [];
    for (const pass of analyzePasses) {
      if (!pass.predicate(currentAst, stability)) {
        continue;
      }
      analyzeResults.push(pass.apply(currentAst, stability));
    }

    // Phase 3 — FOLD: merge all analysis results
    for (const result of analyzeResults) {
      allDiagnostics.push(...result.diagnostics);
      allCertificates.push(...result.certificates);
      if (result.applied) {
        const pass = analyzePasses.find(
          (p) =>
            result.certificates.length > 0 &&
            result.certificates[0]!.passName === p.name
        );
        if (pass) {
          passesApplied.push(pass.name);
        }
      }
    }

    return {
      ast: currentAst,
      diagnostics: allDiagnostics,
      certificates: allCertificates,
      passesApplied,
    };
  }
}

// ─── Default optimizer ───────────────────────────────────────────────

// ─── Self-verification pass (THM-IRREVERSIBILITY-FRAMEWORK) ─────────

class SelfVerificationPass implements OptimizationPass {
  readonly name = 'self-verification';
  readonly theoremId = 'THM-IRREVERSIBILITY-FRAMEWORK';
  readonly priority = 40;
  readonly kind = 'analyze' as const;

  predicate(ast: GraphAST): boolean {
    return ast.edges.some((e) => e.properties.verify !== undefined);
  }

  apply(ast: GraphAST): OptimizationPassResult {
    const annotations: { edge: ASTEdge; verify: string }[] = [];
    for (const edge of ast.edges) {
      if (edge.properties.verify) {
        annotations.push({ edge, verify: edge.properties.verify });
      }
    }

    return {
      ast,
      diagnostics: [],
      certificates:
        annotations.length > 0
          ? [
              {
                passName: this.name,
                theoremId: this.theoremId,
                leanTheoremName: 'self_verification_collected',
                summary: `Collected ${
                  annotations.length
                } verify annotation(s): ${annotations
                  .map((a) => a.verify)
                  .join(', ')}`,
                data: {
                  annotations: annotations.map((a) => ({
                    type: a.edge.type,
                    verify: a.verify,
                    sources: a.edge.sourceIds,
                    targets: a.edge.targetIds,
                  })),
                },
              },
            ]
          : [],
      applied: annotations.length > 0,
    };
  }
}

/**
 * Creates the Default Optimizer.
 */
export function createDefaultOptimizer(): OptimizationPassManager {
  const manager = new OptimizationPassManager();
  manager.register(new ReynoldsFoldRegimePass());
  manager.register(new CoarseningPass());
  manager.register(new BuleyeanTreeshakePass());
  manager.register(new TheoremBackedEliminationPass());
  manager.register(new CodecRacingPass());
  manager.register(new WarmupEfficiencyPass());
  manager.register(new SelfVerificationPass());
  return manager;
}
