// @a0n/gnosis/src/betty/parse-utils.ts
/**
 * parse-utils.ts -- Shared parsing regexes and functions extracted from BettyCompiler.
 *
 * These are the core lexing primitives used by both Betty (the bootstrap compiler)
 * and Betti (the self-hosted compiler). Extracting them enables Betti's FORK
 * lexer nodes to call the same battle-tested parsing logic.
 */

// ============================================================================
// Regex patterns
// ============================================================================

/** Matches standalone node declarations: (id:Label {props}) */
export const NODE_REGEX =
  /\(([^:)\s|]+)(?:\s*:\s*([^){\s]+))?(?:\s*{([^}]+)})?\)/g;

/** Matches edge declarations: (src)-[:TYPE {props}]->(tgt) */
export const EDGE_REGEX =
  /\(([^)]+?)\)\s*-\[:([A-Z]+)(?:\s*{([^}]+)})?\]->\s*\(([^)]+?)\)(?=\s*-|$|\s*\))/g;

/** Matches key-value property pairs inside { } */
export const PROPERTY_PAIR_REGEX =
  /(\w+)\s*:\s*('[^']*'|"[^"]*"|\[[^\]]*\]|[^,]+)/g;

/** Matches node references inside edge source/target groups */
export const NODE_IN_EDGE_REGEX =
  /([^:)\s|{]+)(?:\s*:\s*([^){\s]+))?(?:\s*{([^}]+)})?/g;

/** Matches imperative keywords that gnosis rejects */
export const IMPERATIVE_KEYWORD_REGEX =
  /\b(function|return|if|while|for|var|let|const)\b/;

// ============================================================================
// Parsed fragment types
// ============================================================================

export interface ParsedNode {
  id: string;
  label: string;
  propertiesRaw: string;
}

export interface ParsedEdge {
  sourceRaw: string;
  edgeType: string;
  propertiesRaw: string;
  targetRaw: string;
}

// ============================================================================
// Parsing functions
// ============================================================================

/**
 * Parse all node declarations from a single line.
 * Only matches lines that do NOT contain edge syntax (-[:).
 */
export function parseNodeDeclarations(line: string): ParsedNode[] {
  if (line.includes('-[:')) return [];

  const nodes: ParsedNode[] = [];
  NODE_REGEX.lastIndex = 0;
  let match;
  while ((match = NODE_REGEX.exec(line)) !== null) {
    const id = (match[1] ?? '').trim();
    if (!id || id.includes('|')) continue;
    nodes.push({
      id,
      label: match[2] ? match[2].trim() : '',
      propertiesRaw: match[3] ? match[3].trim() : '',
    });
  }
  return nodes;
}

/**
 * Parse all edge declarations from a single line.
 */
export function parseEdgeDeclarations(line: string): ParsedEdge[] {
  const edges: ParsedEdge[] = [];
  EDGE_REGEX.lastIndex = 0;
  let match;
  while ((match = EDGE_REGEX.exec(line)) !== null) {
    const targetRaw = (match[4] ?? '').trim();
    edges.push({
      sourceRaw: (match[1] ?? '').trim(),
      edgeType: (match[2] ?? '').trim(),
      propertiesRaw: match[3] ? match[3].trim() : '',
      targetRaw,
    });
    // For chained edges like (a)->(b)->(c), reset lastIndex to the start of (b)
    // match.index is the start of (a). We want to start searching again at (b).
    // The targetRaw is wrapped in ().
    const targetStart = line.indexOf(`(${targetRaw})`, match.index);
    if (targetStart >= 0) {
      EDGE_REGEX.lastIndex = targetStart;
    }
  }
  return edges;
}

/**
 * Parse a raw property string like `key: "value", other: 42` into a record.
 */
export function parseProperties(propertiesRaw: string): Record<string, string> {
  if (!propertiesRaw) return {};

  const properties: Record<string, string> = {};
  const pairs = propertiesRaw.match(PROPERTY_PAIR_REGEX);
  if (!pairs) return properties;

  for (const pair of pairs) {
    const separator = pair.indexOf(':');
    if (separator < 0) continue;

    const key = pair.slice(0, separator).trim();
    const value = pair
      .slice(separator + 1)
      .trim()
      .replace(/^['"]|['"]$/g, '');

    if (key.length > 0 && value.length > 0) {
      properties[key] = value;
    }
  }

  return properties;
}

/**
 * Parse node references from an edge's source or target group.
 * Handles piped groups like "a | b | c" and extracts id, label, properties.
 */
export function parseNodeRefsFromEdgeGroup(group: string): ParsedNode[] {
  return group.split('|').map((segment) => {
    NODE_IN_EDGE_REGEX.lastIndex = 0;
    const nm = NODE_IN_EDGE_REGEX.exec(segment.trim());
    if (!nm) return { id: segment.trim(), label: '', propertiesRaw: '' };
    return {
      id: (nm[1] ?? '').trim(),
      label: nm[2] ? nm[2].trim().replace(/[)\s{}]+$/, '') : '',
      propertiesRaw: nm[3] ? nm[3].trim() : '',
    };
  });
}

/**
 * Strip single-line comments from gnosis source.
 * Removes lines that are only comments and strips trailing comments.
 */
export function stripComments(source: string): string {
  return source
    .split('\n')
    .map((line) => line.trim())
    .filter((line) => line && !line.startsWith('//') && !line.startsWith('--') && !line.startsWith('#'))
    .join('\n');
}
