// @a0n/gnosis/src/betty/stability.ts
import type {
  ASTEdge,
  ASTNode,
  Diagnostic,
  DiagnosticCode,
  GraphAST,
} from './compiler';

const THERMODYNAMIC_NODE_LABELS = new Set(['SOURCE', 'STATE', 'SINK']);
const THERMODYNAMIC_PROPERTIES = new Set([
  'pressure',
  'potential',
  'capacity',
  'beta1_target',
  'weight',
  'service_rate',
  'drift_gamma',
  'drift_coefficient',
  'repair_debt',
  'supremum_bound',
  'observable',
  'observable_kind',
  'observable_scale',
  'observable_offset',
  'drift_gap',
  'lyapunov_template',
  'lyapunov_degree',
  'lyapunov_coefficients',
  'state_space',
  'small_set_boundary',
  'minorization_epsilon',
]);
const FORK_EDGE_TYPES = new Set(['FORK', 'EVOLVE', 'SUPERPOSE', 'ENTANGLE']);
const COLLAPSE_EDGE_TYPES = new Set(['FOLD', 'COLLAPSE', 'OBSERVE']);
const VENT_EDGE_TYPES = new Set(['VENT', 'TUNNEL']);
const RACE_EDGE_TYPES = new Set(['RACE']);
const PROCESS_EDGE_TYPES = new Set([
  'PROCESS',
  'EVOLVE',
  'SUPERPOSE',
  'ENTANGLE',
]);
const DEFAULT_SMALL_SOURCE_PRESSURE = 1;
const POWER_ITERATION_STEPS = 24;
const EPSILON = 1e-6;

export type StabilityProofKind =
  | 'none'
  | 'numeric'
  | 'symbolic-reneging'
  | 'bounded-supremum'
  | 'unsupported';

export type LyapunovTemplate =
  | 'affine'
  | 'quadratic'
  | 'polynomial'
  | 'log-barrier'
  | 'piecewise-linear';
export type StateSpaceKind =
  | 'countable'
  | 'continuous-positive'
  | 'continuous-bounded'
  | 'continuous-unbounded';

export interface LyapunovSynthesis {
  template: LyapunovTemplate;
  degree: number;
  coefficients: number[];
  expression: string;
  leanExpression: string;
  driftBound: string;
}

export interface SmallSetDiscovery {
  kind:
    | 'queue-boundary'
    | 'level-set'
    | 'vent-boundary'
    | 'capacity-sublevel'
    | 'compact-sublevel';
  boundary: number;
  expression: string;
  leanSetExpression: string;
  derivedFrom: string;
}

export interface MinorizationSynthesis {
  epsilon: number;
  measureKind: 'dirac' | 'uniform-on-small-set' | 'lebesgue-restricted';
  leanMeasureExpression: string;
  derivedFrom: string;
}

export interface ProductLyapunovComponent {
  stateNodeId: string;
  weight: number;
  lyapunovSynthesis: LyapunovSynthesis;
  smallSetDiscovery: SmallSetDiscovery;
  driftGap: number;
}

export interface ProductLyapunovWitness {
  components: ProductLyapunovComponent[];
  productDriftGap: number;
  productSmallSetExpression: string;
  productLyapunovExpression: string;
  leanProductLyapunovExpression: string;
  leanProductSmallSetExpression: string;
  theoremName: string;
}

export interface StabilityKernelEdge {
  sourceId: string;
  targetId: string;
  edgeType: string;
  rawWeight: number;
  normalizedWeight: number;
  effectiveWeight: number;
}

export interface StabilityStateAssessment {
  nodeId: string;
  potential: string;
  arrival: string | null;
  service: string | null;
  vent: string | null;
  gamma: string | null;
  driftExpression: string;
  status: 'stable' | 'unstable' | 'symbolic' | 'unknown';
}

export interface StabilityProofAssumption {
  name: string;
  leanType: string;
  description: string;
}

export interface StabilityProofObligation {
  theoremName: string;
  kind: StabilityProofKind;
  assumptions: StabilityProofAssumption[];
  summary: string;
  requiresLean: boolean;
}

export interface StabilityRecurrenceStep {
  nodeId: string;
  distanceToSmallSet: number | null;
  nextNodeId: string | null;
}

export interface StabilityRecurrenceWitness {
  finiteStateCertified: boolean;
  steps: StabilityRecurrenceStep[];
}

export interface StabilityCountableQueueWitness {
  stateNodeId: string;
  potential: string;
  queueBoundary: number;
  laminarAtom: number;
  arrivalExpression: string;
  serviceExpression: string;
  ventExpression: string;
  predecessorStepMode: 'margin-predecessor';
  smallSetKind: 'queue-boundary';
}

export type StabilityContinuousObservableKind =
  | 'queue-depth'
  | 'fluid-backlog'
  | 'fractional-retry'
  | 'thermal-load'
  | 'custom-potential';

export interface StabilityContinuousHarrisWitness {
  stateNodeId: string;
  observableKind: StabilityContinuousObservableKind;
  observableExpression: string;
  lyapunovExpression: string;
  expectedObservableExpression: string;
  driftGapExpression: string;
  observableScale: number;
  observableOffset: number;
  driftGap: number;
  smallSetKind: 'queue-boundary' | 'level-set';
  smallSetBoundary: number;
  observableDriftTheoremName: string;
  continuousHarrisTheoremName: string;
  stateSpaceKind?: StateSpaceKind;
  lyapunovSynthesis?: LyapunovSynthesis;
  smallSetDiscovery?: SmallSetDiscovery;
  minorizationSynthesis?: MinorizationSynthesis;
}

export type StabilityHeteroMoAFabricLayerKind = 'cpu' | 'gpu' | 'npu' | 'wasm';

export interface StabilityHeteroMoAFabricLayerWitness {
  kind: StabilityHeteroMoAFabricLayerKind;
  schedulerNodeId: string;
  foldNodeId: string;
  laneCount: number;
}

export interface StabilityHeteroMoAFabricWitness {
  fabricNodeId: string;
  metaSchedulerNodeId: string | null;
  globalCollapseNodeId: string | null;
  cpuLaneCount: number;
  gpuLaneCount: number;
  npuLaneCount: number;
  wasmLaneCount: number;
  totalLaneCount: number;
  activeLayerCount: number;
  pairCount: number;
  mirroredKernelCount: number;
  pairedKernel: string;
  scheduleStrategy: string;
  launchGate: string;
  hedgeDelayTicks: number;
  hedgePolicy: string;
  frameProtocol: string;
  layers: StabilityHeteroMoAFabricLayerWitness[];
  loweringTheoremName: string;
  cannonTheoremName: string;
  pairTheoremName: string;
  wasteTheoremName: string;
  coupledTheoremName: string;
}

export interface StabilityMetadata {
  redline: number | null;
  geometricCeiling: number | null;
  spectralRadius: number | null;
  supremumBound: number | null;
  smallSetNodeIds: string[];
  restorativeGamma: string | null;
  finiteStateRecurrent: boolean;
  proofKind: StabilityProofKind;
  theoremName: string;
  countableQueueCertified: boolean;
  laminarGeometricTheoremName: string | null;
  measurableHarrisTheoremName: string | null;
  measurableLaminarTheoremName: string | null;
  measurableQuantitativeLaminarTheoremName: string | null;
  measurableQuantitativeHarrisTheoremName: string | null;
  measurableFiniteTimeHarrisTheoremName: string | null;
  measurableHarrisRecurrentTheoremName: string | null;
  measurableFiniteTimeGeometricErgodicTheoremName: string | null;
  measurableLevyProkhorovGeometricErgodicTheoremName: string | null;
  measurableLevyProkhorovGeometricDecayTheoremName: string | null;
  measurableLevyProkhorovAbstractGeometricErgodicTheoremName: string | null;
  measurableWitnessQuantitativeHarrisTheoremName: string | null;
  queueBoundary: number | null;
  laminarAtom: number | null;
  queuePotential: string | null;
  continuousHarris: StabilityContinuousHarrisWitness | null;
  productLyapunov: ProductLyapunovWitness | null;
  heteroMoAFabrics?: StabilityHeteroMoAFabricWitness[];
}

export interface StabilityReport {
  enabled: boolean;
  kernelEdges: StabilityKernelEdge[];
  spectralRadius: number | null;
  supremumBound: number | null;
  stateAssessments: StabilityStateAssessment[];
  smallSetNodeIds: string[];
  countableQueue: StabilityCountableQueueWitness | null;
  continuousHarris: StabilityContinuousHarrisWitness | null;
  productLyapunov: ProductLyapunovWitness | null;
  recurrence: StabilityRecurrenceWitness;
  harrisRecurrent: boolean;
  ventIsolationOk: boolean;
  redline: number | null;
  geometricCeiling: number | null;
  restorativeGamma: string | null;
  proof: StabilityProofObligation;
  metadata: StabilityMetadata;
  diagnostics: Diagnostic[];
}

interface ExpandedEdgeCandidate {
  sourceId: string;
  targetId: string;
  edgeType: string;
  rawWeight: number;
  properties: Record<string, string>;
}

interface RateSummary {
  service: number;
  vent: number;
  gamma: number | null;
  serviceRaw: string | null;
  ventRaw: string | null;
  gammaRaw: string | null;
  supremumBound: number | null;
}

function round3(value: number): number {
  return Math.round(value * 1000) / 1000;
}

function sanitizeIdentifier(raw: string): string {
  return raw
    .trim()
    .replace(/[^A-Za-z0-9_]+/g, '_')
    .replace(/^_+|_+$/g, '')
    .toLowerCase();
}

function buildThermalTheoremName(ast: GraphAST): string {
  const sinkNode = [...ast.nodes.values()].find((node) =>
    node.labels.some((label) => label.toUpperCase() === 'SINK')
  );
  const anchor = sinkNode?.id ?? [...ast.nodes.keys()][0] ?? 'gnosis_pipeline';
  const normalized = sanitizeIdentifier(anchor);
  return `${normalized || 'gnosis_pipeline'}_is_geometrically_stable`;
}

function parseFinite(raw: string | undefined): number | null {
  if (!raw) {
    return null;
  }
  const parsed = Number(raw);
  return Number.isFinite(parsed) ? parsed : null;
}

function parseNonnegativeInt(
  raw: string | undefined,
  fallback: number
): number {
  const parsed = parseFinite(raw);
  if (parsed === null || parsed < 0) {
    return fallback;
  }
  return Math.trunc(parsed);
}

function parseHeteroMoAFabricLayerCounts(
  node: ASTNode
): Record<StabilityHeteroMoAFabricLayerKind, number> {
  const inferredCounts: Record<StabilityHeteroMoAFabricLayerKind, number> = {
    cpu: 0,
    gpu: 0,
    npu: 0,
    wasm: 0,
  };

  for (const layer of (node.properties.backend_layers ?? '')
    .split(',')
    .map((value) => value.trim().toLowerCase())
    .filter(
      (value): value is StabilityHeteroMoAFabricLayerKind =>
        value === 'cpu' ||
        value === 'gpu' ||
        value === 'npu' ||
        value === 'wasm'
    )) {
    inferredCounts[layer] += 1;
  }

  return {
    cpu: parseNonnegativeInt(node.properties.cpu_lanes, inferredCounts.cpu),
    gpu: parseNonnegativeInt(node.properties.gpu_lanes, inferredCounts.gpu),
    npu: parseNonnegativeInt(node.properties.npu_lanes, inferredCounts.npu),
    wasm: parseNonnegativeInt(node.properties.wasm_lanes, inferredCounts.wasm),
  };
}

function buildHeteroMoAFabricWitnesses(
  ast: GraphAST,
  theoremName: string
): StabilityHeteroMoAFabricWitness[] {
  const fabricNodes = [...ast.nodes.values()].filter(
    (node) =>
      (node.properties.primitive === 'HeteroMoAFabric' ||
        Boolean(node.properties.backend_layers) ||
        Boolean(node.properties.cpu_lanes) ||
        Boolean(node.properties.gpu_lanes) ||
        Boolean(node.properties.npu_lanes) ||
        Boolean(node.properties.wasm_lanes)) &&
      !node.id.includes('__')
  );

  return fabricNodes.map((node) => {
    const inferredLaneCounts = parseHeteroMoAFabricLayerCounts(node);
    const cpuLaneCount = inferredLaneCounts.cpu;
    const gpuLaneCount = inferredLaneCounts.gpu;
    const npuLaneCount = inferredLaneCounts.npu;
    const wasmLaneCount = inferredLaneCounts.wasm;
    const activeLayerCount = parseNonnegativeInt(
      node.properties.active_layers,
      [cpuLaneCount, gpuLaneCount, npuLaneCount, wasmLaneCount].filter(
        (laneCount) => laneCount > 0
      ).length
    );
    const totalLaneCount =
      cpuLaneCount + gpuLaneCount + npuLaneCount + wasmLaneCount;
    const pairNodes = [...ast.nodes.values()].filter(
      (candidate) =>
        candidate.id.startsWith(`${node.id}__`) &&
        candidate.properties.role === 'paired-kernel-adjudicator'
    );
    const mirroredKernelCount = [...ast.nodes.values()].filter(
      (candidate) =>
        candidate.id.startsWith(`${node.id}__`) &&
        (candidate.properties.pair_role === 'primary' ||
          candidate.properties.pair_role === 'shadow')
    ).length;
    const normalizedFabricId = sanitizeIdentifier(node.id) || 'fabric';
    const theoremPrefix = `${theoremName}_${normalizedFabricId}_moa_fabric`;
    const layers: StabilityHeteroMoAFabricLayerWitness[] = (
      [
        {
          kind: 'cpu' as const,
          laneCount: cpuLaneCount,
        },
        {
          kind: 'gpu' as const,
          laneCount: gpuLaneCount,
        },
        {
          kind: 'npu' as const,
          laneCount: npuLaneCount,
        },
        {
          kind: 'wasm' as const,
          laneCount: wasmLaneCount,
        },
      ] satisfies Array<{
        kind: StabilityHeteroMoAFabricLayerKind;
        laneCount: number;
      }>
    )
      .filter((layer) => layer.laneCount > 0)
      .map((layer) => ({
        ...layer,
        schedulerNodeId: `${node.id}__${layer.kind}__scheduler`,
        foldNodeId: `${node.id}__${layer.kind}__fold`,
      }));

    return {
      fabricNodeId: node.id,
      metaSchedulerNodeId: ast.nodes.has(`${node.id}__meta_scheduler`)
        ? `${node.id}__meta_scheduler`
        : null,
      globalCollapseNodeId: ast.nodes.has(`${node.id}__global_collapse`)
        ? `${node.id}__global_collapse`
        : null,
      cpuLaneCount,
      gpuLaneCount,
      npuLaneCount,
      wasmLaneCount,
      totalLaneCount,
      activeLayerCount,
      pairCount: pairNodes.length > 0 ? pairNodes.length : totalLaneCount,
      mirroredKernelCount:
        mirroredKernelCount > 0 ? mirroredKernelCount : totalLaneCount * 2,
      pairedKernel: node.properties.paired_kernel ?? 'mirror',
      scheduleStrategy: node.properties.schedule_strategy ?? 'cannon',
      launchGate: node.properties.launch_gate ?? 'armed',
      hedgeDelayTicks: parseNonnegativeInt(
        node.properties.hedge_delay_ticks,
        1
      ),
      hedgePolicy: node.properties.hedge_policy ?? 'race',
      frameProtocol: node.properties.frame_protocol ?? 'aeon-10-byte-binary',
      layers,
      loweringTheoremName: `${theoremPrefix}_lowering`,
      cannonTheoremName: `${theoremPrefix}_cannon`,
      pairTheoremName: `${theoremPrefix}_pair`,
      wasteTheoremName: `${theoremPrefix}_waste`,
      coupledTheoremName: `${theoremPrefix}_coupled`,
    };
  });
}

function createDiagnostic(
  code: DiagnosticCode,
  message: string,
  severity: Diagnostic['severity']
): Diagnostic {
  return {
    line: 1,
    column: 1,
    code,
    message,
    severity,
  };
}

function shouldAnalyzeThermodynamics(ast: GraphAST): boolean {
  for (const node of ast.nodes.values()) {
    if (
      node.labels.some((label) =>
        THERMODYNAMIC_NODE_LABELS.has(label.toUpperCase())
      )
    ) {
      return true;
    }

    if (
      Object.keys(node.properties).some((key) =>
        THERMODYNAMIC_PROPERTIES.has(key)
      )
    ) {
      return true;
    }
  }

  for (const edge of ast.edges) {
    if (
      Object.keys(edge.properties).some((key) =>
        THERMODYNAMIC_PROPERTIES.has(key)
      )
    ) {
      return true;
    }
  }

  return false;
}

function buildOutgoingEdges(ast: GraphAST): Map<string, ASTEdge[]> {
  const outgoing = new Map<string, ASTEdge[]>();

  for (const nodeId of ast.nodes.keys()) {
    outgoing.set(nodeId, []);
  }

  for (const edge of ast.edges) {
    for (const sourceId of edge.sourceIds) {
      const normalizedSourceId = sourceId.trim();
      const sourceEdges = outgoing.get(normalizedSourceId) ?? [];
      sourceEdges.push(edge);
      outgoing.set(normalizedSourceId, sourceEdges);
    }
  }

  return outgoing;
}

function buildIncomingEdges(ast: GraphAST): Map<string, ASTEdge[]> {
  const incoming = new Map<string, ASTEdge[]>();

  for (const nodeId of ast.nodes.keys()) {
    incoming.set(nodeId, []);
  }

  for (const edge of ast.edges) {
    for (const targetId of edge.targetIds) {
      const normalizedTargetId = targetId.trim();
      const targetEdges = incoming.get(normalizedTargetId) ?? [];
      targetEdges.push(edge);
      incoming.set(normalizedTargetId, targetEdges);
    }
  }

  return incoming;
}

function defaultRawWeight(edge: ASTEdge): number {
  const explicit = parseFinite(edge.properties.weight);
  if (explicit !== null && explicit >= 0) {
    return explicit;
  }

  if (
    FORK_EDGE_TYPES.has(edge.type.toUpperCase()) ||
    RACE_EDGE_TYPES.has(edge.type.toUpperCase())
  ) {
    return edge.targetIds.length > 0 ? 1 / edge.targetIds.length : 1;
  }

  return 1;
}

function expandEdges(ast: GraphAST): ExpandedEdgeCandidate[] {
  const expanded: ExpandedEdgeCandidate[] = [];

  for (const edge of ast.edges) {
    const rawWeight = defaultRawWeight(edge);
    for (const sourceId of edge.sourceIds) {
      for (const targetId of edge.targetIds) {
        expanded.push({
          sourceId: sourceId.trim(),
          targetId: targetId.trim(),
          edgeType: edge.type.toUpperCase(),
          rawWeight,
          properties: edge.properties,
        });
      }
    }
  }

  return expanded;
}

function buildNormalizedKernel(
  expanded: ExpandedEdgeCandidate[]
): StabilityKernelEdge[] {
  const rawTotals = new Map<string, number>();

  for (const edge of expanded) {
    rawTotals.set(
      edge.sourceId,
      (rawTotals.get(edge.sourceId) ?? 0) + edge.rawWeight
    );
  }

  return expanded.map((edge) => {
    const total = rawTotals.get(edge.sourceId) ?? edge.rawWeight;
    const normalizedWeight = total > 0 ? edge.rawWeight / total : 0;
    return {
      sourceId: edge.sourceId,
      targetId: edge.targetId,
      edgeType: edge.edgeType,
      rawWeight: edge.rawWeight,
      normalizedWeight,
      effectiveWeight: normalizedWeight,
    };
  });
}

function collectSourceNodeIds(ast: GraphAST): string[] {
  const ids = [...ast.nodes.values()]
    .filter(
      (node) =>
        node.labels.some((label) => label.toUpperCase() === 'SOURCE') ||
        typeof node.properties.pressure === 'string'
    )
    .map((node) => node.id);

  return ids.length > 0 ? ids : [...ast.nodes.keys()].slice(0, 1);
}

function basePressureForNode(node: ASTNode): number {
  const pressure = parseFinite(node.properties.pressure);
  if (pressure !== null) {
    return pressure;
  }

  return node.labels.some((label) => label.toUpperCase() === 'SOURCE')
    ? typeof node.properties.pressure === 'string'
      ? 0
      : DEFAULT_SMALL_SOURCE_PRESSURE
    : 0;
}

function propagatePressure(
  ast: GraphAST,
  kernelEdges: StabilityKernelEdge[]
): Map<string, number> {
  const nodeIds = [...ast.nodes.keys()];
  const indexByNodeId = new Map<string, number>(
    nodeIds.map((nodeId, index) => [nodeId, index])
  );
  const base = nodeIds.map((nodeId) => {
    const node = ast.nodes.get(nodeId);
    return node ? basePressureForNode(node) : 0;
  });
  const hasNumericPressure = base.some((value) => value > 0);

  if (!hasNumericPressure) {
    const sourceNodeIds = collectSourceNodeIds(ast);
    for (const sourceNodeId of sourceNodeIds) {
      const sourceIndex = indexByNodeId.get(sourceNodeId);
      if (sourceIndex !== undefined) {
        const sourceNode = ast.nodes.get(sourceNodeId);
        if (!sourceNode || typeof sourceNode.properties.pressure === 'string') {
          continue;
        }
        base[sourceIndex] = DEFAULT_SMALL_SOURCE_PRESSURE;
      }
    }
  }

  let current = [...base];
  for (let step = 0; step < POWER_ITERATION_STEPS; step += 1) {
    const next = [...base];
    for (const edge of kernelEdges) {
      const sourceIndex = indexByNodeId.get(edge.sourceId);
      const targetIndex = indexByNodeId.get(edge.targetId);
      if (sourceIndex === undefined || targetIndex === undefined) {
        continue;
      }
      next[targetIndex] = (next[targetIndex] ?? 0) + (current[sourceIndex] ?? 0) * edge.normalizedWeight;
    }
    current = next;
  }

  return new Map<string, number>(
    nodeIds.map((nodeId, index) => [nodeId, round3(current[index] ?? 0)])
  );
}

function sinkCapacityForTargets(
  ast: GraphAST,
  targetIds: readonly string[]
): number {
  return targetIds.reduce((sum, targetId) => {
    const target = ast.nodes.get(targetId.trim());
    if (!target) {
      return sum;
    }

    const capacity = parseFinite(target.properties.capacity);
    return capacity !== null ? sum + capacity : sum;
  }, 0);
}

function summarizeRates(
  ast: GraphAST,
  outgoingEdges: readonly ASTEdge[]
): RateSummary {
  let service = 0;
  let vent = 0;
  let gamma: number | null = null;
  let serviceRaw: string | null = null;
  let ventRaw: string | null = null;
  let gammaRaw: string | null = null;
  let supremumBound: number | null = null;

  for (const edge of outgoingEdges) {
    const edgeType = edge.type.toUpperCase();
    const edgeGamma = parseFinite(edge.properties.drift_gamma);
    if (edgeGamma !== null && edgeGamma > 0) {
      gamma = gamma === null ? edgeGamma : Math.min(gamma, edgeGamma);
      gammaRaw = edge.properties.drift_gamma ?? null;
    }

    if (RACE_EDGE_TYPES.has(edgeType)) {
      const bound = parseFinite(edge.properties.supremum_bound);
      if (bound !== null) {
        supremumBound =
          supremumBound === null ? bound : Math.min(supremumBound, bound);
      }
    }

    if (COLLAPSE_EDGE_TYPES.has(edgeType)) {
      const serviceRateRaw = edge.properties.service_rate;
      const numericService =
        parseFinite(serviceRateRaw) ??
        (serviceRateRaw ? null : sinkCapacityForTargets(ast, edge.targetIds));
      if (numericService !== null && numericService > 0) {
        service += numericService;
      }
      if (!serviceRaw && typeof serviceRateRaw === 'string') {
        serviceRaw = serviceRateRaw;
      }
    }

    if (VENT_EDGE_TYPES.has(edgeType)) {
      const driftCoefficientRaw = edge.properties.drift_coefficient;
      const driftGammaRaw = edge.properties.drift_gamma;
      const numericVent =
        parseFinite(driftCoefficientRaw) ??
        parseFinite(driftGammaRaw) ??
        (driftCoefficientRaw || driftGammaRaw
          ? null
          : sinkCapacityForTargets(ast, edge.targetIds));
      if (numericVent !== null && numericVent > 0) {
        vent += numericVent;
      }
      if (!ventRaw) {
        ventRaw = driftCoefficientRaw ?? driftGammaRaw ?? null;
      }
    }
  }

  return {
    service,
    vent,
    gamma,
    serviceRaw,
    ventRaw,
    gammaRaw,
    supremumBound,
  };
}

function computeScalingByNode(
  ast: GraphAST,
  pressureByNodeId: Map<string, number>,
  outgoingByNodeId: Map<string, ASTEdge[]>
): {
  scalingByNodeId: Map<string, number>;
  rateByNodeId: Map<string, RateSummary>;
} {
  const scalingByNodeId = new Map<string, number>();
  const rateByNodeId = new Map<string, RateSummary>();

  for (const nodeId of ast.nodes.keys()) {
    const outgoingEdges = outgoingByNodeId.get(nodeId) ?? [];
    const rates = summarizeRates(ast, outgoingEdges);
    rateByNodeId.set(nodeId, rates);

    const arrival = pressureByNodeId.get(nodeId) ?? 0;
    const restorative = rates.service + rates.vent;
    let scaling = 1;

    if (arrival > 0 && restorative > 0) {
      scaling = arrival / restorative;
    } else if (arrival > 0 && restorative === 0 && outgoingEdges.length > 0) {
      scaling = 1.1;
    }

    if (rates.supremumBound !== null) {
      scaling = Math.min(scaling, rates.supremumBound);
    }

    scalingByNodeId.set(nodeId, round3(scaling));
  }

  return {
    scalingByNodeId,
    rateByNodeId,
  };
}

function materializeEffectiveKernel(
  kernelEdges: readonly StabilityKernelEdge[],
  scalingByNodeId: Map<string, number>
): StabilityKernelEdge[] {
  return kernelEdges.map((edge) => ({
    ...edge,
    effectiveWeight: round3(
      edge.normalizedWeight * (scalingByNodeId.get(edge.sourceId) ?? 1)
    ),
  }));
}

function buildAdjacency(ast: GraphAST): Map<string, string[]> {
  const adjacency = new Map<string, string[]>();

  for (const nodeId of ast.nodes.keys()) {
    adjacency.set(nodeId, []);
  }

  for (const edge of ast.edges) {
    for (const sourceId of edge.sourceIds) {
      const normalizedSourceId = sourceId.trim();
      const outgoing = adjacency.get(normalizedSourceId) ?? [];
      outgoing.push(...edge.targetIds.map((targetId) => targetId.trim()));
      adjacency.set(normalizedSourceId, outgoing);
    }
  }

  return adjacency;
}

function buildEffectiveMatrix(
  nodeIds: readonly string[],
  kernelEdges: readonly StabilityKernelEdge[]
): number[][] {
  const indexByNodeId = new Map<string, number>(
    nodeIds.map((nodeId, index) => [nodeId, index])
  );
  const matrix = nodeIds.map(() => nodeIds.map(() => 0));

  for (const edge of kernelEdges) {
    const sourceIndex = indexByNodeId.get(edge.sourceId);
    const targetIndex = indexByNodeId.get(edge.targetId);
    if (sourceIndex === undefined || targetIndex === undefined) {
      continue;
    }
    matrix[sourceIndex]![targetIndex] = (matrix[sourceIndex]![targetIndex] ?? 0) + edge.effectiveWeight;
  }

  return matrix.map((row) => row.map((value) => round3(value)));
}

function estimateSpectralRadius(matrix: readonly number[][]): number {
  if (matrix.length === 0) {
    return 0;
  }

  let vector = matrix.map(() => 1);
  let norm = 1;

  for (let step = 0; step < POWER_ITERATION_STEPS; step += 1) {
    const next = matrix.map((row) =>
      row.reduce((sum, value, index) => sum + value * (vector[index] ?? 0), 0)
    );
    norm = Math.max(...next.map((value) => Math.abs(value)), 0);
    if (norm <= EPSILON) {
      return 0;
    }
    vector = next.map((value) => value / norm);
  }

  return round3(norm);
}

function findStronglyConnectedComponents(
  nodeIds: readonly string[],
  adjacency: Map<string, string[]>
): string[][] {
  let index = 0;
  const stack: string[] = [];
  const indices = new Map<string, number>();
  const lowLinks = new Map<string, number>();
  const onStack = new Set<string>();
  const components: string[][] = [];

  const strongConnect = (nodeId: string): void => {
    indices.set(nodeId, index);
    lowLinks.set(nodeId, index);
    index += 1;
    stack.push(nodeId);
    onStack.add(nodeId);

    for (const neighborId of adjacency.get(nodeId) ?? []) {
      if (!indices.has(neighborId)) {
        strongConnect(neighborId);
        lowLinks.set(
          nodeId,
          Math.min(lowLinks.get(nodeId) ?? 0, lowLinks.get(neighborId) ?? 0)
        );
      } else if (onStack.has(neighborId)) {
        lowLinks.set(
          nodeId,
          Math.min(lowLinks.get(nodeId) ?? 0, indices.get(neighborId) ?? 0)
        );
      }
    }

    if ((lowLinks.get(nodeId) ?? -1) === (indices.get(nodeId) ?? -2)) {
      const component: string[] = [];
      let current = stack.pop();
      while (current !== undefined) {
        onStack.delete(current);
        component.push(current);
        if (current === nodeId) {
          break;
        }
        current = stack.pop();
      }
      components.push(component);
    }
  };

  for (const nodeId of nodeIds) {
    if (!indices.has(nodeId)) {
      strongConnect(nodeId);
    }
  }

  return components;
}

function componentHasCycle(
  component: readonly string[],
  adjacency: Map<string, string[]>
): boolean {
  if (component.length > 1) {
    return true;
  }

  const [nodeId] = component;
  if (nodeId === undefined) return false;
  return (adjacency.get(nodeId) ?? []).includes(nodeId);
}

function canReachSink(
  startNodeId: string,
  sinkNodeIds: Set<string>,
  adjacency: Map<string, string[]>
): boolean {
  const visited = new Set<string>();
  const frontier = [startNodeId];

  while (frontier.length > 0) {
    const current = frontier.pop();
    if (!current || visited.has(current)) {
      continue;
    }

    if (sinkNodeIds.has(current)) {
      return true;
    }

    visited.add(current);
    frontier.push(...(adjacency.get(current) ?? []));
  }

  return false;
}

function buildKernelSupportGraph(
  nodeIds: readonly string[],
  kernelEdges: readonly StabilityKernelEdge[]
): {
  outgoingByNodeId: Map<string, string[]>;
  incomingByNodeId: Map<string, string[]>;
} {
  const outgoingByNodeId = new Map<string, Set<string>>();
  const incomingByNodeId = new Map<string, Set<string>>();

  for (const nodeId of nodeIds) {
    outgoingByNodeId.set(nodeId, new Set());
    incomingByNodeId.set(nodeId, new Set());
  }

  for (const edge of kernelEdges) {
    if (edge.effectiveWeight <= EPSILON) {
      continue;
    }

    outgoingByNodeId.get(edge.sourceId)?.add(edge.targetId);
    incomingByNodeId.get(edge.targetId)?.add(edge.sourceId);
  }

  return {
    outgoingByNodeId: new Map(
      [...outgoingByNodeId.entries()].map(([nodeId, targets]) => [
        nodeId,
        [...targets],
      ])
    ),
    incomingByNodeId: new Map(
      [...incomingByNodeId.entries()].map(([nodeId, sources]) => [
        nodeId,
        [...sources],
      ])
    ),
  };
}

function buildFiniteStateRecurrenceWitness(
  nodeIds: readonly string[],
  kernelEdges: readonly StabilityKernelEdge[],
  sinkNodeIds: readonly string[]
): StabilityRecurrenceWitness {
  if (sinkNodeIds.length === 0) {
    return {
      finiteStateCertified: false,
      steps: nodeIds.map((nodeId) => ({
        nodeId,
        distanceToSmallSet: null,
        nextNodeId: null,
      })),
    };
  }

  const { outgoingByNodeId, incomingByNodeId } = buildKernelSupportGraph(
    nodeIds,
    kernelEdges
  );
  const distanceByNodeId = new Map<string, number>();
  const frontier = [...new Set(sinkNodeIds)];

  for (const sinkNodeId of frontier) {
    distanceByNodeId.set(sinkNodeId, 0);
  }

  while (frontier.length > 0) {
    const currentNodeId = frontier.shift();
    if (!currentNodeId) {
      continue;
    }

    const currentDistance = distanceByNodeId.get(currentNodeId);
    if (currentDistance === undefined) {
      continue;
    }

    for (const sourceNodeId of incomingByNodeId.get(currentNodeId) ?? []) {
      if (distanceByNodeId.has(sourceNodeId)) {
        continue;
      }
      distanceByNodeId.set(sourceNodeId, currentDistance + 1);
      frontier.push(sourceNodeId);
    }
  }

  const steps = nodeIds.map<StabilityRecurrenceStep>((nodeId) => {
    const distanceToSmallSet = distanceByNodeId.get(nodeId) ?? null;
    if (distanceToSmallSet === null || distanceToSmallSet === 0) {
      return {
        nodeId,
        distanceToSmallSet,
        nextNodeId: null,
      };
    }

    const nextNodeId =
      (outgoingByNodeId.get(nodeId) ?? []).find(
        (targetNodeId) =>
          (distanceByNodeId.get(targetNodeId) ?? Number.POSITIVE_INFINITY) ===
          distanceToSmallSet - 1
      ) ?? null;

    return {
      nodeId,
      distanceToSmallSet,
      nextNodeId,
    };
  });

  return {
    finiteStateCertified: steps.every(
      (step) =>
        step.distanceToSmallSet !== null &&
        (step.distanceToSmallSet === 0 || step.nextNodeId !== null)
    ),
    steps,
  };
}

function resolveArrivalExpression(
  nodeId: string,
  incomingByNodeId: Map<string, ASTEdge[]>,
  pressureByNodeId: Map<string, number>,
  ast: GraphAST
): string | null {
  const numericArrival = pressureByNodeId.get(nodeId);
  if (numericArrival !== undefined && numericArrival > 0) {
    return `${numericArrival}`;
  }

  for (const edge of incomingByNodeId.get(nodeId) ?? []) {
    for (const sourceId of edge.sourceIds) {
      const sourceNode = ast.nodes.get(sourceId.trim());
      if (!sourceNode) {
        continue;
      }
      if (typeof sourceNode.properties.pressure === 'string') {
        return sourceNode.properties.pressure;
      }
    }
  }

  return null;
}

function assessStateDrift(
  ast: GraphAST,
  pressureByNodeId: Map<string, number>,
  outgoingByNodeId: Map<string, ASTEdge[]>,
  incomingByNodeId: Map<string, ASTEdge[]>,
  sinkNodeIds: Set<string>,
  rateByNodeId: Map<string, RateSummary>
): {
  assessments: StabilityStateAssessment[];
  diagnostics: Diagnostic[];
  proofKind: StabilityProofKind;
  proofAssumptions: StabilityProofAssumption[];
  restorativeGamma: string | null;
} {
  const diagnostics: Diagnostic[] = [];
  const assessments: StabilityStateAssessment[] = [];
  let proofKind: StabilityProofKind = 'none';
  const proofAssumptions: StabilityProofAssumption[] = [];
  let restorativeGamma: string | null = null;

  const stateNodes = [...ast.nodes.values()].filter(
    (node) =>
      node.labels.some((label) => label.toUpperCase() === 'STATE') ||
      typeof node.properties.potential === 'string'
  );

  for (const stateNode of stateNodes) {
    const rates = rateByNodeId.get(stateNode.id) ?? {
      service: 0,
      vent: 0,
      gamma: null,
      serviceRaw: null,
      ventRaw: null,
      gammaRaw: null,
      supremumBound: null,
    };
    const arrivalExpression = resolveArrivalExpression(
      stateNode.id,
      incomingByNodeId,
      pressureByNodeId,
      ast
    );
    const potential = stateNode.properties.potential ?? 'beta1';
    const serviceExpression =
      rates.serviceRaw ?? (rates.service > 0 ? `${rates.service}` : null);
    const ventExpression =
      rates.ventRaw ?? (rates.vent > 0 ? `${rates.vent}` : null);
    const gammaExpression =
      rates.gammaRaw ?? (rates.gamma !== null ? `${rates.gamma}` : null);
    const arrivalValue = parseFinite(arrivalExpression ?? undefined);
    const serviceValue =
      parseFinite(serviceExpression ?? undefined) ?? rates.service;
    const ventValue = parseFinite(ventExpression ?? undefined) ?? rates.vent;
    const gammaValue =
      parseFinite(gammaExpression ?? undefined) ?? rates.gamma ?? 0;
    const driftExpression = `${arrivalExpression ?? '0'} - ( ${
      serviceExpression ?? '0'
    } + ${ventExpression ?? '0'} )`;

    let status: StabilityStateAssessment['status'] = 'unknown';
    const hasReachableSink = canReachSink(
      stateNode.id,
      sinkNodeIds,
      buildAdjacency(ast)
    );

    if (
      arrivalValue !== null &&
      Number.isFinite(serviceValue) &&
      Number.isFinite(ventValue)
    ) {
      const netDrift = arrivalValue - (serviceValue + ventValue);
      const gammaFloor = gammaValue > 0 ? -gammaValue : 0;
      status =
        netDrift < 0 && netDrift <= gammaFloor + EPSILON
          ? 'stable'
          : 'unstable';

      if (!restorativeGamma && gammaExpression) {
        restorativeGamma = gammaExpression;
      }

      if (status === 'stable') {
        proofKind = proofKind === 'none' ? 'numeric' : proofKind;
      } else {
        diagnostics.push(
          createDiagnostic(
            'ERR_DRIFT_POSITIVE',
            `State '${stateNode.id}' has positive drift (${round3(
              netDrift
            )}). Restorative flow ${round3(
              serviceValue + ventValue
            )} does not dominate pressure ${round3(arrivalValue)}.`,
            'error'
          )
        );
      }
    } else {
      const hasVentEdge = (outgoingByNodeId.get(stateNode.id) ?? []).some(
        (edge) => VENT_EDGE_TYPES.has(edge.type.toUpperCase())
      );
      const repairDebtLeak = (outgoingByNodeId.get(stateNode.id) ?? []).some(
        (edge) => {
          if (!VENT_EDGE_TYPES.has(edge.type.toUpperCase())) {
            return false;
          }
          return (parseFinite(edge.properties.repair_debt) ?? 0) > 0;
        }
      );

      if (
        arrivalExpression &&
        serviceExpression &&
        ventExpression &&
        hasVentEdge &&
        hasReachableSink &&
        !repairDebtLeak
      ) {
        status = 'symbolic';
        proofKind = 'symbolic-reneging';
        restorativeGamma = gammaExpression ?? restorativeGamma;
        if (proofAssumptions.length === 0) {
          proofAssumptions.push({
            name: 'h_drift_floor',
            leanType: 'forall n : Nat, lam - (mu + alpha n) <= -1',
            description:
              'Arrival pressure never exceeds the combined fold and vent floor.',
          });
        }
      } else {
        diagnostics.push(
          createDiagnostic(
            'ERR_DRIFT_POSITIVE',
            `State '${
              stateNode.id
            }' declares potential '${potential}' but Betti cannot prove negative drift from pressure='${
              arrivalExpression ?? 'unknown'
            }', service='${serviceExpression ?? 'unknown'}', vent='${
              ventExpression ?? 'unknown'
            }'.`,
            'error'
          )
        );
      }
    }

    assessments.push({
      nodeId: stateNode.id,
      potential,
      arrival: arrivalExpression,
      service: serviceExpression,
      vent: ventExpression,
      gamma: gammaExpression,
      driftExpression,
      status,
    });
  }

  if (proofKind === 'none' && assessments.length > 0) {
    proofKind = 'unsupported';
  }

  return {
    assessments,
    diagnostics,
    proofKind,
    proofAssumptions,
    restorativeGamma,
  };
}

function collectBranchReachability(
  startNodeId: string,
  outgoingByNodeId: Map<string, ASTEdge[]>,
  sinkNodeIds: Set<string>
): Set<string> {
  const visited = new Set<string>();
  const frontier = [startNodeId];

  while (frontier.length > 0) {
    const current = frontier.pop();
    if (!current || visited.has(current)) {
      continue;
    }
    visited.add(current);

    if (sinkNodeIds.has(current)) {
      continue;
    }

    for (const edge of outgoingByNodeId.get(current) ?? []) {
      const edgeType = edge.type.toUpperCase();
      const isConvergenceBarrier =
        (COLLAPSE_EDGE_TYPES.has(edgeType) ||
          edgeType === 'RACE' ||
          edgeType === 'SLIVER') &&
        edge.sourceIds.length > 1;
      if (isConvergenceBarrier) {
        continue;
      }
      for (const targetId of edge.targetIds) {
        frontier.push(targetId.trim());
      }
    }
  }

  return visited;
}

function assessVentIsolation(
  ast: GraphAST,
  outgoingByNodeId: Map<string, ASTEdge[]>,
  sinkNodeIds: Set<string>
): { ok: boolean; diagnostics: Diagnostic[] } {
  const diagnostics: Diagnostic[] = [];

  for (const edge of ast.edges) {
    const edgeType = edge.type.toUpperCase();
    if (!FORK_EDGE_TYPES.has(edgeType) || edge.targetIds.length < 2) {
      continue;
    }

    const branchMaps = edge.targetIds.map((targetId) =>
      collectBranchReachability(targetId.trim(), outgoingByNodeId, sinkNodeIds)
    );
    const branchOwnerByNodeId = new Map<string, number>();

    branchMaps.forEach((reachableNodeIds, branchIndex) => {
      reachableNodeIds.forEach((nodeId) => {
        const priorBranchIndex = branchOwnerByNodeId.get(nodeId);
        if (
          priorBranchIndex !== undefined &&
          priorBranchIndex !== branchIndex &&
          !sinkNodeIds.has(nodeId)
        ) {
          diagnostics.push(
            createDiagnostic(
              'ERR_REPAIR_DEBT_LEAK',
              `Fork '${edge.sourceIds.join(
                '|'
              )}' loses branch isolation: node '${nodeId}' is shared by sibling branches before a sink/fold barrier.`,
              'error'
            )
          );
        } else {
          branchOwnerByNodeId.set(nodeId, branchIndex);
        }
      });
    });
  }

  for (const edge of ast.edges) {
    const edgeType = edge.type.toUpperCase();
    if (!VENT_EDGE_TYPES.has(edgeType)) {
      continue;
    }

    const repairDebt = parseFinite(edge.properties.repair_debt) ?? 0;
    if (repairDebt > 0) {
      diagnostics.push(
        createDiagnostic(
          'ERR_REPAIR_DEBT_LEAK',
          `Vent '${edge.sourceIds.join('|')} -> ${edge.targetIds.join(
            '|'
          )}' carries repair debt ${repairDebt}. Vent paths must remain zero-debt under branch isolation.`,
          'error'
        )
      );
    }
  }

  return {
    ok: diagnostics.length === 0,
    diagnostics,
  };
}

function computeRedline(
  ast: GraphAST,
  sinkNodeIds: readonly string[],
  spectralRadius: number | null
): number | null {
  const numericCapacities = sinkNodeIds
    .map((sinkNodeId) =>
      parseFinite(ast.nodes.get(sinkNodeId)?.properties.capacity)
    )
    .filter((capacity): capacity is number => capacity !== null);

  if (numericCapacities.length > 0) {
    return Math.min(...numericCapacities);
  }

  if (spectralRadius !== null && spectralRadius < 1) {
    return round3(1 / Math.max(1 - spectralRadius, 0.001));
  }

  return null;
}

function buildCountableQueueWitness(
  proofKind: StabilityProofKind,
  assessments: readonly StabilityStateAssessment[],
  redline: number | null
): StabilityCountableQueueWitness | null {
  if (proofKind !== 'symbolic-reneging' && proofKind !== 'numeric') {
    return null;
  }

  const queueAssessment = assessments.find(
    (assessment) =>
      assessment.status === 'symbolic' || assessment.status === 'stable'
  );
  if (!queueAssessment?.arrival || !queueAssessment.service) {
    return null;
  }

  return {
    stateNodeId: queueAssessment.nodeId,
    potential: queueAssessment.potential,
    queueBoundary:
      typeof redline === 'number' && Number.isFinite(redline) && redline >= 0
        ? Math.trunc(redline)
        : 0,
    laminarAtom: 0,
    arrivalExpression: queueAssessment.arrival,
    serviceExpression: queueAssessment.service,
    ventExpression: queueAssessment.vent ?? '0',
    predecessorStepMode: 'margin-predecessor',
    smallSetKind: 'queue-boundary',
  };
}

function normalizeObservableKind(
  raw: string | undefined
): StabilityContinuousObservableKind | null {
  switch (raw?.trim().toLowerCase()) {
    case 'queue-depth':
      return 'queue-depth';
    case 'fluid-backlog':
      return 'fluid-backlog';
    case 'fractional-retry':
      return 'fractional-retry';
    case 'thermal-load':
      return 'thermal-load';
    case 'custom-potential':
      return 'custom-potential';
    default:
      return null;
  }
}

function formatObservableNumber(value: number): string {
  const rounded = round3(value);
  return `${rounded}`;
}

function formatAffineObservableExpression(
  baseLabel: string,
  scale: number,
  offset: number
): string {
  const normalizedBaseLabel = baseLabel.trim() || 'queue_depth';
  const normalizedScale = round3(scale);
  const normalizedOffset = round3(offset);
  const scaledTerm =
    Math.abs(normalizedScale - 1) <= EPSILON
      ? normalizedBaseLabel
      : `${formatObservableNumber(normalizedScale)} * ${normalizedBaseLabel}`;

  if (Math.abs(normalizedOffset) <= EPSILON) {
    return scaledTerm;
  }

  if (normalizedOffset > 0) {
    return `${scaledTerm} + ${formatObservableNumber(normalizedOffset)}`;
  }

  return `${scaledTerm} - ${formatObservableNumber(
    Math.abs(normalizedOffset)
  )}`;
}

function buildObservableBaseLabel(
  stateNode: ASTNode | undefined,
  potential: string,
  observableKind: StabilityContinuousObservableKind
): string {
  const explicitObservable = stateNode?.properties.observable?.trim();
  if (explicitObservable) {
    return explicitObservable;
  }

  switch (observableKind) {
    case 'fluid-backlog':
      return 'fluid_backlog';
    case 'fractional-retry':
      return 'fractional_retry';
    case 'thermal-load':
      return 'thermal_load';
    case 'custom-potential':
    case 'queue-depth':
      return potential;
  }
}

function inferContinuousObservableKind(
  stateNode: ASTNode | undefined,
  potential: string
): StabilityContinuousObservableKind {
  const explicitKind = normalizeObservableKind(
    stateNode?.properties.observable_kind
  );
  if (explicitKind) {
    return explicitKind;
  }

  const descriptor = [
    stateNode?.properties.observable,
    stateNode?.properties.lyapunov,
    potential,
  ]
    .filter((value): value is string => typeof value === 'string')
    .join(' ')
    .toLowerCase();

  if (
    descriptor.includes('temperature') ||
    descriptor.includes('thermal') ||
    descriptor.includes('heat')
  ) {
    return 'thermal-load';
  }

  if (descriptor.includes('retry')) {
    return 'fractional-retry';
  }

  if (
    descriptor.includes('fluid') ||
    descriptor.includes('backlog') ||
    descriptor.includes('buffer') ||
    descriptor.includes('pressure')
  ) {
    return 'fluid-backlog';
  }

  if (
    descriptor.includes('queue') ||
    descriptor.includes('depth') ||
    potential.trim().toLowerCase() === 'beta1'
  ) {
    return 'queue-depth';
  }

  return 'custom-potential';
}

function normalizeStateSpaceKind(raw: string | undefined): StateSpaceKind {
  switch (raw?.trim().toLowerCase()) {
    case 'continuous-positive':
      return 'continuous-positive';
    case 'continuous-bounded':
      return 'continuous-bounded';
    case 'continuous-unbounded':
      return 'continuous-unbounded';
    case 'countable':
    default:
      return 'countable';
  }
}

function normalizeLyapunovTemplate(raw: string | undefined): LyapunovTemplate {
  switch (raw?.trim().toLowerCase()) {
    case 'quadratic':
      return 'quadratic';
    case 'polynomial':
      return 'polynomial';
    case 'log-barrier':
      return 'log-barrier';
    case 'piecewise-linear':
      return 'piecewise-linear';
    case 'affine':
    default:
      return 'affine';
  }
}

function parseCoefficients(raw: string | undefined): number[] {
  if (!raw) return [];
  return raw
    .split(',')
    .map((s) => Number(s.trim()))
    .filter((n) => Number.isFinite(n));
}

/**
 * Handles the gnosis synthesize Lyapunov workflow.
 */
export function synthesizeLyapunov(
  stateNode: ASTNode | undefined,
  _observableKind: StabilityContinuousObservableKind,
  template: LyapunovTemplate,
  driftData: { scale: number; offset: number; driftGap: number }
): LyapunovSynthesis {
  const explicitCoeffs = parseCoefficients(
    stateNode?.properties.lyapunov_coefficients
  );
  const scale = driftData.scale;
  const offset = driftData.offset;

  switch (template) {
    case 'affine': {
      const coefficients =
        explicitCoeffs.length >= 2 ? explicitCoeffs : [scale, offset];
      return {
        template: 'affine',
        degree: 1,
        coefficients,
        expression: `${formatObservableNumber(
          coefficients[0] ?? 0
        )} * x + ${formatObservableNumber(coefficients[1] ?? 0)}`,
        leanExpression: `fun x => ${formatObservableNumber(
          coefficients[0] ?? 0
        )} * x + ${formatObservableNumber(coefficients[1] ?? 0)}`,
        driftBound: `-(${formatObservableNumber(driftData.driftGap)})`,
      };
    }
    case 'quadratic': {
      const c = (explicitCoeffs.length >= 1 ? explicitCoeffs[0] : scale) ?? 0;
      const o = (explicitCoeffs.length >= 2 ? explicitCoeffs[1] : offset) ?? 0;
      return {
        template: 'quadratic',
        degree: 2,
        coefficients: [c, o],
        expression: `${formatObservableNumber(
          c
        )} * x^2 + ${formatObservableNumber(o)}`,
        leanExpression: `fun x => ${formatObservableNumber(
          c
        )} * x ^ 2 + ${formatObservableNumber(o)}`,
        driftBound: `-(${formatObservableNumber(driftData.driftGap)})`,
      };
    }
    case 'polynomial': {
      const rawCoeffs =
        explicitCoeffs.length >= 2 ? explicitCoeffs : [scale, 0, offset];
      // Pad to exactly four coefficients for realPolynomialObservable(c0, c1, c2, c3)
      const coeffs = [
        rawCoeffs[0] ?? 0,
        rawCoeffs[1] ?? 0,
        rawCoeffs[2] ?? 0,
        rawCoeffs[3] ?? 0,
      ];
      const terms = coeffs
        .map((c, i) =>
          i === 0
            ? formatObservableNumber(c)
            : `${formatObservableNumber(c)} * x^${i}`
        )
        .filter((_, i) => Math.abs(coeffs[i] ?? 0) > EPSILON || i === 0);
      return {
        template: 'polynomial',
        degree: Math.max(1, rawCoeffs.length - 1),
        coefficients: coeffs,
        expression: terms.join(' + '),
        leanExpression: `fun x => ${formatObservableNumber(
          coeffs[0] ?? 0
        )} + ${formatObservableNumber(
          coeffs[1] ?? 0
        )} * x + ${formatObservableNumber(
          coeffs[2] ?? 0
        )} * x ^ 2 + ${formatObservableNumber(coeffs[3] ?? 0)} * x ^ 3`,
        driftBound: `-(${formatObservableNumber(driftData.driftGap)})`,
      };
    }
    case 'log-barrier': {
      const c = (explicitCoeffs.length >= 1 ? explicitCoeffs[0] : scale) ?? 0;
      return {
        template: 'log-barrier',
        degree: 1,
        coefficients: [c],
        expression: `${formatObservableNumber(c)} * log(1 + x)`,
        leanExpression: `fun x => ${formatObservableNumber(
          c
        )} * Real.log (1 + x)`,
        driftBound: `-(${formatObservableNumber(driftData.driftGap)})`,
      };
    }
    case 'piecewise-linear': {
      const pieces =
        explicitCoeffs.length >= 2 ? explicitCoeffs : [scale, offset];
      const maxTerms: string[] = [];
      for (let i = 0; i < pieces.length; i += 2) {
        const s = pieces[i] ?? 1;
        const o = pieces[i + 1] ?? 0;
        maxTerms.push(
          `${formatObservableNumber(s)} * x + ${formatObservableNumber(o)}`
        );
      }
      return {
        template: 'piecewise-linear',
        degree: 1,
        coefficients: pieces,
        expression: `max(${maxTerms.join(', ')})`,
        leanExpression: `fun x => ${
          maxTerms.length === 1
            ? `${formatObservableNumber(
                pieces[0] ?? 1
              )} * x + ${formatObservableNumber(pieces[1] ?? 0)}`
            : maxTerms.reduce((acc, term) => `max (${acc}) (${term})`)
        }`,
        driftBound: `-(${formatObservableNumber(driftData.driftGap)})`,
      };
    }
  }
}

/**
 * Handles the gnosis discover Small Set workflow.
 */
export function discoverSmallSet(
  ast: GraphAST,
  lyapunovSynthesis: LyapunovSynthesis,
  stateNode: ASTNode | undefined,
  edges: readonly ASTEdge[]
): SmallSetDiscovery {
  // Priority 1: explicit small_set_boundary
  const explicitBoundary = parseFinite(
    stateNode?.properties.small_set_boundary
  );
  if (explicitBoundary !== null && explicitBoundary > 0) {
    return {
      kind: 'level-set',
      boundary: explicitBoundary,
      expression: `{x | V(x) <= ${formatObservableNumber(explicitBoundary)}}`,
      leanSetExpression: `{x | ${lyapunovSynthesis.leanExpression.replace(
        'fun x => ',
        ''
      )} ≤ ${formatObservableNumber(explicitBoundary)}}`,
      derivedFrom: 'explicit small_set_boundary property',
    };
  }

  // Priority 2: VENT boundary inference
  for (const edge of edges) {
    if (!VENT_EDGE_TYPES.has(edge.type.toUpperCase())) continue;
    const condition = edge.properties.condition?.trim();
    if (!condition) continue;
    const match = condition.match(/^\s*\w+\s*>\s*([\d.]+)\s*$/);
    if (match) {
      const ventBoundary = Number(match[1]);
      if (Number.isFinite(ventBoundary) && ventBoundary > 0) {
        return {
          kind: 'vent-boundary',
          boundary: ventBoundary,
          expression: `{x | x <= ${formatObservableNumber(ventBoundary)}}`,
          leanSetExpression: `Set.Iic ${formatObservableNumber(ventBoundary)}`,
          derivedFrom: `VENT condition: ${condition}`,
        };
      }
    }
  }

  // Priority 3: capacity sublevel from downstream SINK
  const sinkNodes = [...ast.nodes.values()].filter((node) =>
    node.labels.some((label) => label.toUpperCase() === 'SINK')
  );
  for (const sink of sinkNodes) {
    const capacity = parseFinite(sink.properties.capacity);
    if (capacity !== null && capacity > 0) {
      return {
        kind: 'capacity-sublevel',
        boundary: capacity,
        expression: `{x | x <= ${formatObservableNumber(capacity)}}`,
        leanSetExpression: `Set.Iic ${formatObservableNumber(capacity)}`,
        derivedFrom: `SINK capacity: ${capacity}`,
      };
    }
  }

  // Priority 4: Lyapunov level set
  const driftGap = parseFinite(stateNode?.properties.drift_gap) ?? 1;
  const levelSetBoundary = Math.max(1, Math.ceil(driftGap * 2));
  return {
    kind: 'compact-sublevel',
    boundary: levelSetBoundary,
    expression: `{x | V(x) <= ${formatObservableNumber(levelSetBoundary)}}`,
    leanSetExpression: `{x | ${lyapunovSynthesis.leanExpression.replace(
      'fun x => ',
      ''
    )} ≤ ${formatObservableNumber(levelSetBoundary)}}`,
    derivedFrom: 'Lyapunov level-set fallback',
  };
}

/**
 * Handles the gnosis synthesize Minorization workflow.
 */
export function synthesizeMinorization(
  stateSpaceKind: StateSpaceKind,
  smallSet: SmallSetDiscovery,
  lyapunovSynthesis: LyapunovSynthesis,
  driftData: { driftGap: number; explicitEpsilon: number | null }
): MinorizationSynthesis {
  // Explicit override
  if (driftData.explicitEpsilon !== null && driftData.explicitEpsilon > 0) {
    return {
      epsilon: driftData.explicitEpsilon,
      measureKind:
        stateSpaceKind === 'countable' ? 'dirac' : 'uniform-on-small-set',
      leanMeasureExpression:
        stateSpaceKind === 'countable'
          ? 'MeasureTheory.Measure.dirac 0'
          : `(1 / ${formatObservableNumber(
              smallSet.boundary
            )}) • MeasureTheory.Measure.restrict MeasureTheory.volume (${
              smallSet.leanSetExpression
            })`,
      derivedFrom: 'explicit minorization_epsilon property',
    };
  }

  switch (stateSpaceKind) {
    case 'countable':
      return {
        epsilon: 1,
        measureKind: 'dirac',
        leanMeasureExpression: 'MeasureTheory.Measure.dirac 0',
        derivedFrom: 'countable queue: dirac at atom',
      };
    case 'continuous-positive': {
      const lyapMaxOnC =
        (lyapunovSynthesis.coefficients[0] ?? 0) * smallSet.boundary +
        (lyapunovSynthesis.coefficients[1] ?? 0);
      const epsilon =
        lyapMaxOnC > 0 ? round3(driftData.driftGap / (2 * lyapMaxOnC)) : 0.5;
      return {
        epsilon: Math.max(EPSILON, epsilon),
        measureKind: 'uniform-on-small-set',
        leanMeasureExpression: `(1 / ${formatObservableNumber(
          smallSet.boundary
        )}) • MeasureTheory.Measure.restrict MeasureTheory.volume (${
          smallSet.leanSetExpression
        })`,
        derivedFrom: 'continuous-positive: uniform on small set',
      };
    }
    case 'continuous-bounded': {
      const invCapacity = smallSet.boundary > 0 ? 1 / smallSet.boundary : 0.5;
      const epsilon = round3(Math.min(driftData.driftGap, invCapacity));
      return {
        epsilon: Math.max(EPSILON, epsilon),
        measureKind: 'lebesgue-restricted',
        leanMeasureExpression: `(1 / ${formatObservableNumber(
          smallSet.boundary
        )}) • MeasureTheory.Measure.restrict MeasureTheory.volume (${
          smallSet.leanSetExpression
        })`,
        derivedFrom: 'continuous-bounded: Lebesgue restricted to small set',
      };
    }
    case 'continuous-unbounded': {
      const lyapMaxOnC =
        (lyapunovSynthesis.coefficients[0] ?? 0) * smallSet.boundary +
        (lyapunovSynthesis.coefficients[1] ?? 0);
      const epsilon =
        lyapMaxOnC > 0 ? round3(driftData.driftGap / (2 * lyapMaxOnC)) : 0.5;
      return {
        epsilon: Math.max(EPSILON, epsilon),
        measureKind: 'uniform-on-small-set',
        leanMeasureExpression: `(1 / ${formatObservableNumber(
          smallSet.boundary
        )}) • MeasureTheory.Measure.restrict MeasureTheory.volume (${
          smallSet.leanSetExpression
        })`,
        derivedFrom: 'continuous-unbounded: uniform on small set',
      };
    }
  }
}

function buildProductLyapunovWitness(
  ast: GraphAST,
  _continuousHarris: StabilityContinuousHarrisWitness | null,
  theoremName: string
): ProductLyapunovWitness | null {
  // Find all continuous state nodes (non-countable state_space with synthesis data)
  const continuousStateNodes = [...ast.nodes.values()].filter((node) => {
    const ssKind = normalizeStateSpaceKind(node.properties.state_space);
    return ssKind !== 'countable' && node.properties.drift_gap;
  });

  if (continuousStateNodes.length < 2) {
    return null;
  }

  // Check for PROCESS edges connecting continuous state nodes
  const continuousNodeIds = new Set(continuousStateNodes.map((n) => n.id));
  const hasProcessEdge = ast.edges.some((edge) => {
    if (!PROCESS_EDGE_TYPES.has(edge.type.toUpperCase())) return false;
    const sourceMatch = edge.sourceIds.some((id) =>
      continuousNodeIds.has(id.trim())
    );
    const targetMatch = edge.targetIds.some((id) =>
      continuousNodeIds.has(id.trim())
    );
    return sourceMatch && targetMatch;
  });

  // Also detect implicit coupling via shared SINK/FOLD targets
  const hasCoupledSink = (() => {
    const sinkTargets = new Map<string, string[]>();
    for (const edge of ast.edges) {
      if (!COLLAPSE_EDGE_TYPES.has(edge.type.toUpperCase())) continue;
      for (const targetId of edge.targetIds) {
        for (const sourceId of edge.sourceIds) {
          if (continuousNodeIds.has(sourceId.trim())) {
            const existing = sinkTargets.get(targetId.trim()) ?? [];
            existing.push(sourceId.trim());
            sinkTargets.set(targetId.trim(), existing);
          }
        }
      }
    }
    return [...sinkTargets.values()].some((sources) => sources.length >= 2);
  })();

  if (!hasProcessEdge && !hasCoupledSink) {
    return null;
  }

  // Synthesize per-component Lyapunov data
  const components: ProductLyapunovComponent[] = [];
  for (const node of continuousStateNodes) {
    const template = normalizeLyapunovTemplate(
      node.properties.lyapunov_template
    );
    const obsKind =
      normalizeObservableKind(node.properties.observable_kind) ??
      inferContinuousObservableKind(node, node.properties.potential ?? 'x');
    const scale = parseFinite(node.properties.observable_scale) ?? 1;
    const offset = parseFinite(node.properties.observable_offset) ?? 0;
    const dg = parseFinite(node.properties.drift_gap) ?? 1;
    if (dg <= 0 || scale <= 0) continue;

    const lyapSynthesis = synthesizeLyapunov(node, obsKind, template, {
      scale,
      offset,
      driftGap: dg,
    });
    const smallSetDisc = discoverSmallSet(ast, lyapSynthesis, node, ast.edges);

    components.push({
      stateNodeId: node.id,
      weight: 1, // equal weighting by default
      lyapunovSynthesis: lyapSynthesis,
      smallSetDiscovery: smallSetDisc,
      driftGap: dg,
    });
  }

  if (components.length < 2) {
    return null;
  }

  const productGap = Math.min(...components.map((c) => c.weight * c.driftGap));
  const productSmallSetParts = components.map(
    (c) =>
      `{${c.stateNodeId} | V(${c.stateNodeId}) <= ${formatObservableNumber(
        c.smallSetDiscovery.boundary
      )}}`
  );
  const productLyapParts = components.map(
    (c) =>
      `${formatObservableNumber(c.weight)} * V_${c.stateNodeId}(${
        c.stateNodeId
      })`
  );
  const leanProductLyapParts = components.map(
    (c) =>
      `${formatObservableNumber(
        c.weight
      )} * (${c.lyapunovSynthesis.leanExpression.replace('fun x => ', '')})`
  );

  return {
    components,
    productDriftGap: round3(productGap),
    productSmallSetExpression: productSmallSetParts.join(' × '),
    productLyapunovExpression: productLyapParts.join(' + '),
    leanProductLyapunovExpression: `fun p => ${leanProductLyapParts
      .join(' + ')
      .replace(/x/g, 'p.1')}`,
    leanProductSmallSetExpression: `productSmallSet (${components[0]!.smallSetDiscovery.leanSetExpression}) (${components[1]!.smallSetDiscovery.leanSetExpression})`,
    theoremName: `${theoremName}_product_lyapunov`,
  };
}

function buildContinuousHarrisWitness(
  ast: GraphAST,
  countableQueue: StabilityCountableQueueWitness | null,
  theoremName: string
): {
  witness: StabilityContinuousHarrisWitness | null;
  diagnostics: Diagnostic[];
} {
  const diagnostics: Diagnostic[] = [];

  // If no countable queue witness, try direct continuous synthesis
  // for nodes that declare observable_kind explicitly
  if (!countableQueue) {
    const continuousStateNode = [...ast.nodes.values()].find(
      (node) =>
        node.properties.observable_kind &&
        [
          'fluid-backlog',
          'fractional-retry',
          'thermal-load',
          'custom-potential',
        ].includes(
          normalizeObservableKind(node.properties.observable_kind) ?? ''
        ) &&
        node.properties.drift_gap
    );
    if (!continuousStateNode) {
      return { witness: null, diagnostics };
    }
    const obsKind = normalizeObservableKind(
      continuousStateNode.properties.observable_kind
    )!;
    const scale =
      parseFinite(continuousStateNode.properties.observable_scale) ?? 1;
    const offset =
      parseFinite(continuousStateNode.properties.observable_offset) ?? 0;
    const driftGap = parseFinite(continuousStateNode.properties.drift_gap);
    if (!driftGap || driftGap <= 0 || scale <= 0) {
      return { witness: null, diagnostics };
    }

    const stateSpaceKind = normalizeStateSpaceKind(
      continuousStateNode.properties.state_space
    );
    const template = normalizeLyapunovTemplate(
      continuousStateNode.properties.lyapunov_template
    );

    // Use new synthesis pipeline for non-countable state spaces
    if (stateSpaceKind !== 'countable') {
      const lyapSynthesis = synthesizeLyapunov(
        continuousStateNode,
        obsKind,
        template,
        { scale, offset, driftGap }
      );
      const smallSetDisc = discoverSmallSet(
        ast,
        lyapSynthesis,
        continuousStateNode,
        ast.edges
      );
      const minSynthesis = synthesizeMinorization(
        stateSpaceKind,
        smallSetDisc,
        lyapSynthesis,
        {
          driftGap,
          explicitEpsilon: parseFinite(
            continuousStateNode.properties.minorization_epsilon
          ),
        }
      );
      const baseLabel = buildObservableBaseLabel(
        continuousStateNode,
        continuousStateNode.properties.potential ?? 'x',
        obsKind
      );
      const obsExpr = formatAffineObservableExpression(
        baseLabel,
        scale,
        offset
      );
      const expectedObsExpr = formatAffineObservableExpression(
        baseLabel,
        scale,
        offset - scale
      );
      return {
        witness: {
          stateNodeId: continuousStateNode.id,
          observableKind: obsKind,
          observableExpression: obsExpr,
          lyapunovExpression: lyapSynthesis.expression,
          expectedObservableExpression: expectedObsExpr,
          driftGapExpression: formatObservableNumber(driftGap),
          observableScale: round3(scale),
          observableOffset: round3(offset),
          driftGap: round3(driftGap),
          smallSetKind: 'level-set',
          smallSetBoundary: smallSetDisc.boundary,
          observableDriftTheoremName: `${theoremName}_measurable_observable_drift`,
          continuousHarrisTheoremName: `${theoremName}_measurable_continuous_harris_certified`,
          stateSpaceKind,
          lyapunovSynthesis: lyapSynthesis,
          smallSetDiscovery: smallSetDisc,
          minorizationSynthesis: minSynthesis,
        },
        diagnostics,
      };
    }

    const baseLabel = buildObservableBaseLabel(
      continuousStateNode,
      continuousStateNode.properties.potential ?? 'x',
      obsKind
    );
    const obsExpr = formatAffineObservableExpression(baseLabel, scale, offset);
    const expectedObsExpr = formatAffineObservableExpression(
      baseLabel,
      scale,
      offset - scale
    );
    const smallSetBoundary = Math.max(1, Math.ceil(offset / driftGap));
    return {
      witness: {
        stateNodeId: continuousStateNode.id,
        observableKind: obsKind,
        observableExpression: obsExpr,
        lyapunovExpression: obsExpr,
        expectedObservableExpression: expectedObsExpr,
        driftGapExpression: formatObservableNumber(driftGap),
        observableScale: round3(scale),
        observableOffset: round3(offset),
        driftGap: round3(driftGap),
        smallSetKind: 'level-set',
        smallSetBoundary,
        observableDriftTheoremName: `${theoremName}_measurable_observable_drift`,
        continuousHarrisTheoremName: `${theoremName}_measurable_continuous_harris_certified`,
      },
      diagnostics,
    };
  }

  const stateNode = ast.nodes.get(countableQueue.stateNodeId);

  // Non-countable state space declared on a node with a countable queue witness:
  // use the continuous synthesis pipeline instead of the affine queue path.
  const queueStateSpaceKind = normalizeStateSpaceKind(
    stateNode?.properties.state_space
  );
  if (queueStateSpaceKind !== 'countable' && stateNode) {
    const obsKind =
      normalizeObservableKind(stateNode.properties.observable_kind) ??
      inferContinuousObservableKind(stateNode, countableQueue.potential);
    if (obsKind) {
      const scale = parseFinite(stateNode.properties.observable_scale) ?? 1;
      const offset = parseFinite(stateNode.properties.observable_offset) ?? 0;
      const dg = parseFinite(stateNode.properties.drift_gap) ?? scale;
      if (dg > 0 && scale > 0) {
        const template = normalizeLyapunovTemplate(
          stateNode.properties.lyapunov_template
        );
        const lyapSynthesis = synthesizeLyapunov(stateNode, obsKind, template, {
          scale,
          offset,
          driftGap: dg,
        });
        const smallSetDisc = discoverSmallSet(
          ast,
          lyapSynthesis,
          stateNode,
          ast.edges
        );
        const minSynthesis = synthesizeMinorization(
          queueStateSpaceKind,
          smallSetDisc,
          lyapSynthesis,
          {
            driftGap: dg,
            explicitEpsilon: parseFinite(
              stateNode.properties.minorization_epsilon
            ),
          }
        );
        const baseLabel = buildObservableBaseLabel(
          stateNode,
          countableQueue.potential,
          obsKind
        );
        const obsExpr = formatAffineObservableExpression(
          baseLabel,
          scale,
          offset
        );
        const expectedObsExpr = formatAffineObservableExpression(
          baseLabel,
          scale,
          offset - scale
        );
        return {
          witness: {
            stateNodeId: countableQueue.stateNodeId,
            observableKind: obsKind,
            observableExpression: obsExpr,
            lyapunovExpression: lyapSynthesis.expression,
            expectedObservableExpression: expectedObsExpr,
            driftGapExpression: formatObservableNumber(dg),
            observableScale: round3(scale),
            observableOffset: round3(offset),
            driftGap: round3(dg),
            smallSetKind: 'level-set',
            smallSetBoundary: smallSetDisc.boundary,
            observableDriftTheoremName: `${theoremName}_measurable_observable_drift`,
            continuousHarrisTheoremName: `${theoremName}_measurable_continuous_harris_certified`,
            stateSpaceKind: queueStateSpaceKind,
            lyapunovSynthesis: lyapSynthesis,
            smallSetDiscovery: smallSetDisc,
            minorizationSynthesis: minSynthesis,
          },
          diagnostics,
        };
      }
    }
  }

  const rawObservableKind = stateNode?.properties.observable_kind;
  const observableKind = rawObservableKind
    ? normalizeObservableKind(rawObservableKind)
    : inferContinuousObservableKind(stateNode, countableQueue.potential);
  if (!observableKind) {
    diagnostics.push(
      createDiagnostic(
        'ERR_CONTINUOUS_WITNESS_INVALID',
        `State '${countableQueue.stateNodeId}' declares unsupported observable_kind='${rawObservableKind}'. Betti currently supports queue-depth, fluid-backlog, fractional-retry, thermal-load, and custom-potential.`,
        'error'
      )
    );
    return {
      witness: null,
      diagnostics,
    };
  }

  const rawScale = stateNode?.properties.observable_scale;
  const observableScale = rawScale === undefined ? 1 : parseFinite(rawScale);
  if (observableScale === null || observableScale <= 0) {
    diagnostics.push(
      createDiagnostic(
        'ERR_CONTINUOUS_WITNESS_INVALID',
        `State '${
          countableQueue.stateNodeId
        }' requires observable_scale to be a positive number, got '${
          rawScale ?? 'undefined'
        }'.`,
        'error'
      )
    );
    return {
      witness: null,
      diagnostics,
    };
  }

  const rawOffset = stateNode?.properties.observable_offset;
  const observableOffset = rawOffset === undefined ? 0 : parseFinite(rawOffset);
  if (observableOffset === null) {
    diagnostics.push(
      createDiagnostic(
        'ERR_CONTINUOUS_WITNESS_INVALID',
        `State '${countableQueue.stateNodeId}' requires observable_offset to be numeric, got '${rawOffset}'.`,
        'error'
      )
    );
    return {
      witness: null,
      diagnostics,
    };
  }

  const rawDriftGap = stateNode?.properties.drift_gap;
  const explicitDriftGap =
    rawDriftGap === undefined ? null : parseFinite(rawDriftGap);
  if (
    rawDriftGap !== undefined &&
    (explicitDriftGap === null || explicitDriftGap <= 0)
  ) {
    diagnostics.push(
      createDiagnostic(
        'ERR_CONTINUOUS_WITNESS_INVALID',
        `State '${countableQueue.stateNodeId}' requires drift_gap to be a positive number, got '${rawDriftGap}'.`,
        'error'
      )
    );
    return {
      witness: null,
      diagnostics,
    };
  }

  const driftGap = explicitDriftGap ?? observableScale;
  if (driftGap - observableScale > EPSILON) {
    diagnostics.push(
      createDiagnostic(
        'ERR_CONTINUOUS_WITNESS_INVALID',
        `State '${
          countableQueue.stateNodeId
        }' sets drift_gap='${rawDriftGap}', but the current affine queue witness only supports 0 < drift_gap <= observable_scale = ${formatObservableNumber(
          observableScale
        )}.`,
        'error'
      )
    );
    return {
      witness: null,
      diagnostics,
    };
  }

  const observableBaseLabel = buildObservableBaseLabel(
    stateNode,
    countableQueue.potential,
    observableKind
  );
  const observableExpression = formatAffineObservableExpression(
    observableBaseLabel,
    observableScale,
    observableOffset
  );
  const expectedObservableExpression = formatAffineObservableExpression(
    observableBaseLabel,
    observableScale,
    observableOffset - observableScale
  );

  return {
    witness: {
      stateNodeId: countableQueue.stateNodeId,
      observableKind,
      observableExpression,
      lyapunovExpression: observableExpression,
      expectedObservableExpression,
      driftGapExpression: formatObservableNumber(driftGap),
      observableScale: round3(observableScale),
      observableOffset: round3(observableOffset),
      driftGap: round3(driftGap),
      smallSetKind: countableQueue.smallSetKind,
      smallSetBoundary: countableQueue.queueBoundary,
      observableDriftTheoremName: `${theoremName}_measurable_observable_drift`,
      continuousHarrisTheoremName: `${theoremName}_measurable_continuous_harris_certified`,
    },
    diagnostics,
  };
}

/**
 * Handles the gnosis analyze Topology Stability workflow.
 */
export function analyzeTopologyStability(
  ast: GraphAST,
  currentBettiNumber: number
): StabilityReport | null {
  if (!shouldAnalyzeThermodynamics(ast)) {
    return null;
  }

  const diagnostics: Diagnostic[] = [];
  const nodeIds = [...ast.nodes.keys()];

  // ── FORK: three independent structural analyses ──────────────────
  // Per THM-SERVER-OPTIMALITY, independent computations are forked.
  // These three share no data dependencies.
  const outgoingByNodeId = buildOutgoingEdges(ast);
  const incomingByNodeId = buildIncomingEdges(ast);
  const adjacency = buildAdjacency(ast);
  // ── FOLD: structural analyses complete ───────────────────────────

  // ── PROCESS chain A: pressure → scaling → kernel → matrix → spectral
  const baseKernel = buildNormalizedKernel(expandEdges(ast));
  const pressureByNodeId = propagatePressure(ast, baseKernel);
  const { scalingByNodeId, rateByNodeId } = computeScalingByNode(
    ast,
    pressureByNodeId,
    outgoingByNodeId
  );
  const kernelEdges = materializeEffectiveKernel(baseKernel, scalingByNodeId);
  const matrix = buildEffectiveMatrix(nodeIds, kernelEdges);
  const spectralRadius = estimateSpectralRadius(matrix);
  // ── Chain A complete: spectral radius available ──────────────────
  const sinkNodeIds = [...ast.nodes.values()]
    .filter(
      (node) =>
        node.labels.some((label) => label.toUpperCase() === 'SINK') ||
        parseFinite(node.properties.beta1_target) === 0
    )
    .map((node) => node.id);
  const sinkNodeIdSet = new Set<string>(sinkNodeIds);
  const supremumBound = [...rateByNodeId.values()]
    .map((rates) => rates.supremumBound)
    .filter((bound): bound is number => bound !== null)
    .reduce<number | null>(
      (lowest, bound) => (lowest === null ? bound : Math.min(lowest, bound)),
      null
    );

  if (currentBettiNumber > 0 && sinkNodeIds.length === 0) {
    diagnostics.push(
      createDiagnostic(
        'ERR_BETA_UNBOUNDED',
        `The topology retains beta1=${currentBettiNumber} but declares no Sink/small-set node. Betti cannot prove the covering space returns to zero debt.`,
        'error'
      )
    );
  }

  const components = findStronglyConnectedComponents(nodeIds, adjacency);
  for (const component of components) {
    if (!componentHasCycle(component, adjacency)) {
      continue;
    }

    const cycleScales = component.map(
      (nodeId) => scalingByNodeId.get(nodeId) ?? 1
    );
    const canDrain = component.some((nodeId) =>
      canReachSink(nodeId, sinkNodeIdSet, adjacency)
    );
    if (!canDrain || Math.max(...cycleScales) >= 1) {
      diagnostics.push(
        createDiagnostic(
          'ERR_BETA_UNBOUNDED',
          `Cycle '${component.join(
            ' -> '
          )}' has no stable geometric floor. A feedback loop either cannot reach the small set or amplifies pressure with load factor ${round3(
            Math.max(...cycleScales)
          )}.`,
          'error'
        )
      );
    }
  }

  const threshold = supremumBound ?? 1;
  if (spectralRadius >= threshold - EPSILON) {
    diagnostics.push(
      createDiagnostic(
        'ERR_SPECTRAL_EXPLOSION',
        `Spectral radius ${spectralRadius} exceeds the admissible ceiling ${round3(
          threshold
        )}. The routing kernel cannot dissipate failure flow fast enough.`,
        'error'
      )
    );
  }

  // ── FORK: three independent assessment passes ─────────────────────
  // Drift assessment, vent isolation, and recurrence witness are
  // independent analyses that share inputs but not outputs.
  const driftAssessment = assessStateDrift(
    ast,
    pressureByNodeId,
    outgoingByNodeId,
    incomingByNodeId,
    sinkNodeIdSet,
    rateByNodeId
  );

  const ventIsolation = assessVentIsolation(
    ast,
    outgoingByNodeId,
    sinkNodeIdSet
  );

  const recurrence = buildFiniteStateRecurrenceWitness(
    nodeIds,
    kernelEdges,
    sinkNodeIds
  );
  // ── FOLD: merge assessment diagnostics ───────────────────────────
  diagnostics.push(...driftAssessment.diagnostics);
  diagnostics.push(...ventIsolation.diagnostics);
  const harrisRecurrent = recurrence.finiteStateCertified;

  const hasSpectralNoDriftCertificate =
    driftAssessment.proofKind === 'none' && spectralRadius < 1;
  const proofKind = hasSpectralNoDriftCertificate
    ? 'bounded-supremum'
    : driftAssessment.proofKind;
  const theoremName = buildThermalTheoremName(ast);
  const redline = computeRedline(ast, sinkNodeIds, spectralRadius);
  const geometricCeiling = round3(supremumBound ?? spectralRadius);
  const countableQueue = buildCountableQueueWitness(
    proofKind,
    driftAssessment.assessments,
    redline
  );
  const continuousHarrisResult = buildContinuousHarrisWitness(
    ast,
    countableQueue,
    theoremName
  );
  diagnostics.push(...continuousHarrisResult.diagnostics);
  const continuousHarris = continuousHarrisResult.witness;
  const productLyapunov = buildProductLyapunovWitness(
    ast,
    continuousHarris,
    theoremName
  );
  const proofAssumptions =
    driftAssessment.proofAssumptions.length > 0
      ? driftAssessment.proofAssumptions
      : proofKind === 'numeric'
      ? [
          {
            name: 'h_drift_floor',
            leanType: 'forall n : Nat, lam - (mu + alpha n) <= -1',
            description:
              'Concrete rates already satisfy the negative-drift floor.',
          },
        ]
      : [];
  const proofSummary =
    proofKind === 'symbolic-reneging'
      ? 'Reneging-family queue detected; export the drift floor to Lean and discharge it with derive_gnosis_drift.'
      : proofKind === 'numeric'
      ? 'Numeric thermodynamic bounds are sufficient to certify the drift floor directly.'
      : proofKind === 'bounded-supremum'
      ? supremumBound !== null
        ? 'Adaptive routing is constrained by an explicit supremum bound below one.'
        : 'The routing kernel is spectrally stable below one and carries no drift obligation.'
      : 'No thermodynamic proof family could be synthesized from the annotated topology.';

  const proof: StabilityProofObligation = {
    theoremName,
    kind: proofKind,
    assumptions: proofAssumptions,
    summary: proofSummary,
    requiresLean:
      proofKind === 'symbolic-reneging' || proofKind === 'bounded-supremum',
  };

  const heteroMoAFabrics = buildHeteroMoAFabricWitnesses(ast, theoremName);

  const metadata: StabilityMetadata = {
    redline,
    geometricCeiling,
    spectralRadius,
    supremumBound,
    smallSetNodeIds: sinkNodeIds,
    restorativeGamma: driftAssessment.restorativeGamma,
    finiteStateRecurrent: recurrence.finiteStateCertified,
    proofKind,
    theoremName,
    countableQueueCertified: countableQueue !== null,
    laminarGeometricTheoremName:
      countableQueue !== null
        ? `${theoremName}_laminar_geometric_stable`
        : null,
    measurableHarrisTheoremName:
      countableQueue !== null
        ? `${theoremName}_measurable_harris_certified`
        : null,
    measurableLaminarTheoremName:
      countableQueue !== null
        ? `${theoremName}_measurable_laminar_certified`
        : null,
    measurableQuantitativeLaminarTheoremName:
      countableQueue !== null
        ? `${theoremName}_measurable_quantitative_laminar_certified`
        : null,
    measurableQuantitativeHarrisTheoremName:
      countableQueue !== null
        ? `${theoremName}_measurable_quantitative_harris_certified`
        : null,
    measurableFiniteTimeHarrisTheoremName:
      countableQueue !== null
        ? `${theoremName}_measurable_finite_time_harris_recurrent`
        : null,
    measurableHarrisRecurrentTheoremName:
      countableQueue !== null
        ? `${theoremName}_measurable_harris_recurrent`
        : null,
    measurableFiniteTimeGeometricErgodicTheoremName:
      countableQueue !== null
        ? `${theoremName}_measurable_finite_time_geometric_ergodic`
        : null,
    measurableLevyProkhorovGeometricErgodicTheoremName:
      countableQueue !== null
        ? `${theoremName}_measurable_levy_prokhorov_geometric_ergodic`
        : null,
    measurableLevyProkhorovGeometricDecayTheoremName:
      countableQueue !== null
        ? `${theoremName}_measurable_levy_prokhorov_geometric_decay`
        : null,
    measurableLevyProkhorovAbstractGeometricErgodicTheoremName:
      countableQueue !== null
        ? `${theoremName}_measurable_levy_prokhorov_geometric_ergodic_abstract`
        : null,
    measurableWitnessQuantitativeHarrisTheoremName:
      countableQueue !== null
        ? `${theoremName}_measurable_witness_quantitative_harris_certified`
        : null,
    queueBoundary: countableQueue?.queueBoundary ?? null,
    laminarAtom: countableQueue?.laminarAtom ?? null,
    queuePotential: countableQueue?.potential ?? null,
    continuousHarris,
    productLyapunov,
    heteroMoAFabrics,
  };

  return {
    enabled: true,
    kernelEdges,
    spectralRadius,
    supremumBound,
    stateAssessments: driftAssessment.assessments,
    smallSetNodeIds: sinkNodeIds,
    countableQueue,
    continuousHarris,
    productLyapunov,
    recurrence,
    harrisRecurrent,
    ventIsolationOk: ventIsolation.ok,
    redline,
    geometricCeiling,
    restorativeGamma: driftAssessment.restorativeGamma,
    proof,
    metadata,
    diagnostics,
  };
}
