// @a0n/gnosis/src/betty/verify.ts
/**
 * verify.ts -- Standalone verification passes extracted from BettyCompiler.
 *
 * Each pass is a pure function operating on (ast, b1) or (ast, stability).
 * BettyCompiler.parse() calls runAllVerificationPasses() which orchestrates them.
 * Betti's verifier handler calls the same passes.
 */

import type { GraphAST, Diagnostic, DeficitAnalysis } from './compiler';
import type { StabilityReport } from './stability';
import type { OptimizerResult } from './optimizer';
import type { CoarseningSynthesisResult } from './coarsen';
import type { SemanticCompatibilityResult } from './semantic-compatibility';
import type { HopeCertificate } from './semantic-hope';
import { analyzeTopologyStability } from './stability';
import { createDefaultOptimizer } from './optimizer';
import {
  extractFiberPartition,
  validatePartition,
  synthesizeCoarsening,
} from './coarsen';
import { analyzeSemanticCompatibility } from './semantic-compatibility';
import { generateHopeCertificate } from './semantic-hope';
import {
  checkSevenLaws,
  checkGodFormula,
  checkControlStatistics,
  type CheckResult,
  type BuleyeanInput,
} from '../seven-laws-checker';
import { certifyGnosisAst } from '../law';

// ============================================================================
// Verification result
// ============================================================================

export interface VerificationResult {
  diagnostics: Diagnostic[];
  stability: StabilityReport | null;
  optimizer: OptimizerResult | null;
  coarseningSynthesis: CoarseningSynthesisResult | null;
  voidDimensions: number;
  landauerHeat: number;
  deficit: DeficitAnalysis | null;
  semanticCompatibility: SemanticCompatibilityResult | null;
  hopeCertificate: HopeCertificate | null;
  /** Clockwork self-verification verdict (THM-VOID-BOUNDARY-SUFFICIENT-STATISTIC) */
  clockwork: import('../universe/index.js').ClockworkVerdict;
  /** Universe certificate emitted by the God Formula kernel. */
  universe: import('../universe/index.js').UniverseCertificate;
  /** The AST after optimizer transforms (may differ from input). */
  ast: GraphAST;
}

// ============================================================================
// Individual pass helpers
// ============================================================================

/**
 * Compute void dimensions: sum of FORK branch counts.
 */
export function computeVoidDimensions(ast: GraphAST): number {
  let dims = 0;
  for (const edge of ast.edges) {
    if (edge.type === 'FORK') {
      dims += edge.targetIds.length;
    }
  }
  return dims;
}

/**
 * Compute Landauer heat: kT ln(2) per irreversible bit erasure.
 * Each FOLD/COLLAPSE/OBSERVE merging N sources erases log2(N) bits.
 */
export function computeLandauerHeat(ast: GraphAST): number {
  const kT = 1; // normalized
  let heat = 0;
  for (const edge of ast.edges) {
    if (
      edge.type === 'FOLD' ||
      edge.type === 'COLLAPSE' ||
      edge.type === 'OBSERVE'
    ) {
      const n = Math.max(2, edge.sourceIds.length);
      heat += kT * Math.log2(n);
    } else if (edge.type === 'VENT') {
      heat += kT;
    }
  }
  return heat;
}

/**
 * Compute per-node deficit analysis.
 * Deficit = outgoing branching factor - incoming merge factor.
 */
export function computeDeficitAnalysis(ast: GraphAST): DeficitAnalysis {
  const perNode = new Map<string, number>();
  const diagnostics: Diagnostic[] = [];

  for (const [nodeId] of ast.nodes) {
    let outBranching = 0;
    let inMerging = 0;

    for (const edge of ast.edges) {
      if (edge.sourceIds.includes(nodeId)) {
        outBranching += edge.targetIds.length;
      }
      if (edge.targetIds.includes(nodeId)) {
        inMerging += edge.sourceIds.length;
      }
    }

    const deficit = outBranching - inMerging;
    perNode.set(nodeId, deficit);
  }

  const totalDeficit = Array.from(perNode.values()).reduce(
    (sum, d) => sum + Math.abs(d),
    0
  );

  if (totalDeficit > 0) {
    diagnostics.push({
      line: 1,
      column: 1,
      code: 'INFO_DEFICIT_SUBGRAPH',
      message: `Topology contains ${totalDeficit} unit(s) of deficit across ${perNode.size} node(s).`,
      severity: 'info',
    });
  }

  return { perNode, totalDeficit, diagnostics };
}

// ============================================================================
// Seven Laws & God Formula Verification Pass
// ============================================================================

/** Map from CheckResult id to DiagnosticCode. */
const CHECK_TO_CODE: Record<string, import('./compiler.js').DiagnosticCode> = {
  'law-1': 'LAW_IMPOSSIBILITY_OF_ZERO',
  'law-2': 'LAW_STRICT_ORDERING',
  'law-3': 'LAW_UNIVERSAL_SANDWICH',
  'law-4': 'LAW_CAVE_DEFICIT',
  'law-5': 'LAW_CONSERVATION',
  'law-6': 'LAW_SORITES_SHARPNESS',
  'law-7': 'LAW_CHAIN_TERMINATION',
  'gf-1': 'GOD_FORMULA_EXISTENCE',
  'gf-2': 'GOD_FORMULA_PERSISTENCE',
  'gf-3': 'GOD_FORMULA_LEARNING',
  'gf-4': 'GOD_FORMULA_DISCRIMINATION',
  'gf-5': 'GOD_FORMULA_BOUNDEDNESS',
  'gf-6': 'GOD_FORMULA_CONVERGENCE',
  'gf-7': 'GOD_FORMULA_UNCERTAINTY',
  'gf-8': 'GOD_FORMULA_COHERENCE',
  'gf-hope': 'GOD_FORMULA_HOPE',
  'ctrl-1': 'CTRL_INDIVIDUAL',
  'ctrl-2': 'CTRL_AGGREGATE',
  'ctrl-3': 'CTRL_SPREAD',
  'ctrl-4': 'CTRL_MEAN',
};

/**
 * Extract Buleyean input from the topology's execution structure.
 *
 * The topology's FORK/RACE/FOLD edges define the choice space.
 * Each FOLD merging N sources produces N choices, each VENT is a rejection.
 * The observation round count is the number of FOLD + OBSERVE edges.
 */
function extractBuleyeanInput(ast: GraphAST): BuleyeanInput | null {
  // Count per-node rejections from VENT edges, rounds from FOLD/OBSERVE
  const nodeRejections = new Map<string, number>();
  let totalRounds = 0;

  for (const edge of ast.edges) {
    if (
      edge.type === 'FOLD' ||
      edge.type === 'COLLAPSE' ||
      edge.type === 'OBSERVE'
    ) {
      totalRounds++;
      for (const sid of edge.sourceIds) {
        if (!nodeRejections.has(sid)) nodeRejections.set(sid, 0);
      }
    }
    if (edge.type === 'VENT') {
      for (const sid of edge.sourceIds) {
        nodeRejections.set(sid, (nodeRejections.get(sid) ?? 0) + 1);
      }
    }
    if (edge.type === 'RACE') {
      // All but one racer are implicitly rejected
      for (const sid of edge.sourceIds) {
        if (!nodeRejections.has(sid)) nodeRejections.set(sid, 0);
      }
    }
  }

  if (nodeRejections.size === 0) return null;
  const rejections = Array.from(nodeRejections.values());
  return { rounds: Math.max(totalRounds, 1), rejections };
}

/**
 * Run Seven Laws, God Formula, and Control Statistics checks on a topology.
 * Violations are returned as first-order compiler diagnostics (errors).
 */
export function verifySevenLawsAndGodFormula(ast: GraphAST): Diagnostic[] {
  const diagnostics: Diagnostic[] = [];
  const input = extractBuleyeanInput(ast);
  if (!input) return diagnostics;

  const lawReport = checkSevenLaws(input);
  const gfReport = checkGodFormula(input);
  const ctrlReport = checkControlStatistics(input);

  const allChecks: CheckResult[] = [
    ...lawReport.laws,
    ...gfReport.consequences,
    gfReport.hope,
    ...ctrlReport.controls,
  ];

  for (const check of allChecks) {
    if (!check.pass) {
      const code = CHECK_TO_CODE[check.id];
      diagnostics.push({
        line: 0,
        column: 0,
        ...(code !== undefined && { code }),
        message: `[${check.leanTheorem}] ${check.name}: ${check.message}`,
        severity: 'error',
      });
    }
  }

  // Also emit info-level diagnostics for the control statistics
  if (ctrlReport.pass && input.rejections.length > 0) {
    diagnostics.push({
      line: 0,
      column: 0,
      message: `Buleyean audit: N=${ctrlReport.stats.numChoices} R=${ctrlReport.stats.rounds} W=${ctrlReport.stats.totalWeight} spread=${ctrlReport.stats.maxSpread} -- all 7 laws, 8 consequences, 4 controls pass`,
      severity: 'info',
    });
  }

  return diagnostics;
}

// ============================================================================
// Orchestrator
// ============================================================================

/**
 * Run all verification passes on an AST.
 *
 * This is the standalone equivalent of the verification section in
 * BettyCompiler.parse() (lines 467-594). The BettyCompiler instance
 * methods (checkTaggedNodeExhaustiveness, checkVoidWalkerInvariants, etc.)
 * remain on the class for now -- this function runs the passes that
 * operate on pure data without needing the compiler's internal state.
 */
export function runAllVerificationPasses(
  ast: GraphAST,
  b1: number
): VerificationResult {
  const diagnostics: Diagnostic[] = [];
  let currentAst = ast;

  // ── FORK: independent passes that only need the AST ──────────────
  // These run first because they don't depend on stability/optimizer.
  // Conceptually forked: semantic on one path, metrics on another.

  // Semantic compatibility (independent of stability chain)
  const semanticCompatibility = analyzeSemanticCompatibility(currentAst);
  diagnostics.push(...semanticCompatibility.diagnostics);

  // O(E) metric passes (independent of everything -- just read edges)
  const voidDimensions = computeVoidDimensions(currentAst);
  const landauerHeat = computeLandauerHeat(currentAst);
  const deficit = computeDeficitAnalysis(currentAst);
  diagnostics.push(...deficit.diagnostics);

  // Seven Laws & God Formula verification (independent -- reads topology structure)
  const sevenLawsDiags = verifySevenLawsAndGodFormula(currentAst);
  diagnostics.push(...sevenLawsDiags);

  const universeEvaluation = certifyGnosisAst(currentAst, {
    subject: 'gnosis-topology',
    metadata: {
      verifier: 'betty',
      beta1: b1,
    },
  });
  const clockwork = universeEvaluation.verdict;
  const universe = universeEvaluation.certificate;

  // ── SEQUENTIAL CHAIN: stability → optimizer → coarsening ─────────
  // These form a blocking dependency chain.

  const stability = analyzeTopologyStability(currentAst, b1);
  if (stability) {
    diagnostics.push(...stability.diagnostics);

    // Optimizer passes (needs stability)
    const optimizer = createDefaultOptimizer();
    const optimizerResult = optimizer.apply(currentAst, stability);
    currentAst = optimizerResult.ast;
    diagnostics.push(...optimizerResult.diagnostics);

    // Coarsening synthesis (needs stability + optimized AST)
    const partition = extractFiberPartition(currentAst);
    let coarseningSynthesis: CoarseningSynthesisResult | null = null;
    if (partition) {
      const valDiags = validatePartition(partition, stability);
      diagnostics.push(...valDiags);
      if (!valDiags.some((d) => d.severity === 'error')) {
        coarseningSynthesis = synthesizeCoarsening(
          currentAst,
          stability,
          partition
        );
        diagnostics.push(...coarseningSynthesis.diagnostics);
      }
    }

    // ── FOLD: merge independent + sequential results ─────────────
    // Hope certificate depends on semantic (already done above)
    const hopeCertificate = generateHopeCertificate(
      currentAst,
      semanticCompatibility
    );

    return {
      diagnostics,
      stability,
      optimizer: optimizerResult,
      coarseningSynthesis,
      voidDimensions,
      landauerHeat,
      deficit,
      semanticCompatibility,
      hopeCertificate,
      clockwork,
      universe,
      ast: currentAst,
    };
  }

  // No stability report -- independent passes already computed above
  return {
    diagnostics,
    stability: null,
    optimizer: null,
    coarseningSynthesis: null,
    voidDimensions,
    landauerHeat,
    deficit,
    semanticCompatibility,
    hopeCertificate: null,
    clockwork,
    universe,
    ast: currentAst,
  };
}
