// @a0n/.ignored_gnosis/mesh-local-mcp/src/aeon-monitor-probe.ts
import { constants as fsConstants } from 'node:fs';
import { accessSync } from 'node:fs';
import { resolve } from 'node:path';

/**
 * Handles the gnosis can Execute File workflow.
 */
export function canExecuteFile(path: string): boolean {
  try {
    accessSync(path, fsConstants.X_OK);
    return true;
  } catch {
    return false;
  }
}

/**
 * Mirrors `wrapper-mesh-delegation.mjs` candidate order for `monster-mesh --mesh-backend native`.
 * When `explicitBin` is set (config or resolved override), only that path is considered.
 */
export function aeonMonitorCandidatePaths(gnosisRoot: string, explicitBin: string | null): readonly string[] {
  const trimmed = explicitBin?.trim() ?? '';
  if (trimmed.length > 0) {
    return [resolve(trimmed)];
  }
  return [
    resolve(gnosisRoot, 'distributed-inference/target/release/aeon-monitor'),
    resolve(gnosisRoot, 'target/release/aeon-monitor'),
  ];
}

/**
 * Handles the gnosis find Executable Aeon Monitor workflow.
 */
export function findExecutableAeonMonitor(
  gnosisRoot: string,
  explicitBin: string | null,
): string | null {
  for (const p of aeonMonitorCandidatePaths(gnosisRoot, explicitBin)) {
    if (canExecuteFile(p)) return p;
  }
  return null;
}

export interface AeonMonitorProbeResult {
  readonly gnosisRoot: string;
  /** Effective `GNOSIS_AEON_MONITOR_BIN`-style override (single path) or null to use default search. */
  readonly explicitBin: string | null;
  readonly candidates: ReadonlyArray<{ readonly path: string; readonly executable: boolean }>;
  readonly resolvedExecutable: string | null;
  readonly nativeBackendReady: boolean;
}

/**
 * Handles the gnosis probe Aeon Monitor workflow.
 */
export function probeAeonMonitor(gnosisRoot: string, explicitBin: string | null): AeonMonitorProbeResult {
  const paths = aeonMonitorCandidatePaths(gnosisRoot, explicitBin);
  const candidates = paths.map((path) => ({
    path,
    executable: canExecuteFile(path),
  }));
  const resolved =
    candidates.find((c) => c.executable)?.path ??
    null;
  return {
    gnosisRoot,
    explicitBin: explicitBin?.trim() || null,
    candidates,
    resolvedExecutable: resolved,
    nativeBackendReady: resolved !== null,
  };
}
