// @a0n/.ignored_gnosis/mesh-local-mcp/src/aeon-monitor-build.ts
import { spawn } from 'node:child_process';

export interface AeonMonitorBuildResult {
  readonly ok: boolean;
  readonly code: string;
  readonly exitCode: number | null;
  readonly timedOut: boolean;
  readonly stdout: string;
  readonly stderr: string;
  readonly message: string;
}

const MAX_OUT = 512_000;

function collectStream(child: ReturnType<typeof spawn>, stream: 'stdout' | 'stderr'): Promise<string> {
  return new Promise((resolvePromise, reject) => {
    const chunks: Buffer[] = [];
    const s = child[stream];
    if (!s) {
      resolvePromise('');
      return;
    }
    s.on('data', (c: Buffer) => {
      chunks.push(c);
      const total = chunks.reduce((n, b) => n + b.length, 0);
      if (total > MAX_OUT) {
        chunks.length = 0;
        chunks.push(Buffer.from('…[truncated]'));
      }
    });
    s.on('error', reject);
    s.on('end', () => resolvePromise(Buffer.concat(chunks).toString('utf8')));
  });
}

export interface AeonMonitorBuildOptions {
  readonly timeoutMs?: number;
}

/**
 * Runs `pnpm run aeon-monitor:build` from the Gnosis package root (same as `monster-mesh:build:native-monitor`).
 */
export function buildAeonMonitor(
  gnosisRoot: string,
  options: AeonMonitorBuildOptions = {},
): Promise<AeonMonitorBuildResult> {
  return new Promise((resolvePromise) => {
    const timeoutMs = Math.max(1_000, options.timeoutMs ?? 120_000);
    const child = spawn('pnpm', ['run', 'aeon-monitor:build'], {
      cwd: gnosisRoot,
      stdio: ['ignore', 'pipe', 'pipe'],
      env: process.env,
    });
    let timedOut = false;
    const timeout = setTimeout(() => {
      timedOut = true;
      child.kill('SIGTERM');
    }, timeoutMs);

    const outP = collectStream(child, 'stdout');
    const errP = collectStream(child, 'stderr');

    child.on('error', (err) => {
      void Promise.all([outP, errP]).then(([stdout, stderr]) => {
        clearTimeout(timeout);
        resolvePromise({
          ok: false,
          code: 'SPAWN_FAILED',
          exitCode: null,
          timedOut,
          stdout,
          stderr,
          message: err instanceof Error ? err.message : String(err),
        });
      });
    });

    child.on('close', (code) => {
      void Promise.all([outP, errP]).then(([stdout, stderr]) => {
        clearTimeout(timeout);
        const ok = code === 0;
        const buildCode = timedOut ? 'BUILD_TIMEOUT' : ok ? 'BUILD_OK' : 'BUILD_FAILED';
        resolvePromise({
          ok,
          code: buildCode,
          exitCode: code,
          timedOut,
          stdout,
          stderr,
          message: timedOut
            ? `aeon-monitor build timed out after ${timeoutMs}ms`
            : ok
            ? 'aeon-monitor build finished'
            : `cargo/pnpm exited with code ${code ?? 'null'}`,
        });
      });
    });
  });
}
