// @a0n/.ignored_gnosis/mesh-local-mcp/src/main.ts
/**
 * Local MCP + operator SSE for Gnosis mesh / RTL-SDR flows.
 * MCP transport: HTTP GET /mcp (SSE) + POST /messages (see MCP SDK simpleSseServer).
 * Telemetry: GET /mesh/events (custom JSON in SSE `data:` lines).
 */

import type { Server } from 'node:http';
import { dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { SSEServerTransport } from '@modelcontextprotocol/sdk/server/sse.js';
import { createMcpExpressApp } from '@modelcontextprotocol/sdk/server/express.js';
import type { Application } from 'express';
import { probeAeonMonitor } from './aeon-monitor-probe.js';
import { MeshRuntime } from './mesh-runtime.js';
import { MeshEventFanout } from './sse-fanout.js';
import { registerMeshMcpTools } from './mcp-register.js';

const __dirname = dirname(fileURLToPath(import.meta.url));

function gnosisRootFromPackage(): string {
  return resolve(__dirname, '..', '..');
}

export interface MeshLocalMcpListenOptions {
  readonly host?: string;
  readonly port?: number;
}

export interface MeshLocalMcpHandle {
  readonly baseUrl: string;
  readonly app: Application;
  close(): Promise<void>;
}

function meshMcpServer(runtime: MeshRuntime, fanout: MeshEventFanout): McpServer {
  const server = new McpServer(
    { name: 'gnosis-mesh-local-mcp', version: '0.1.0' },
    { capabilities: { tools: {} } },
  );
  registerMeshMcpTools(server, runtime, fanout);
  return server;
}

/**
 * Creates the Mesh Local Mcp App.
 */
export function createMeshLocalMcpApp(
  runtime: MeshRuntime,
  fanout: MeshEventFanout,
  host: string,
): { app: Application; transports: Record<string, SSEServerTransport> } {
  const app = createMcpExpressApp({ host });
  const transports: Record<string, SSEServerTransport> = {};

  app.get('/health', (_req, res) => {
    res.json({ ok: true, service: 'gnosis-mesh-local-mcp' });
  });

  app.get('/mesh/status', (req, res) => {
    const rawLimit = Number.parseInt(String(req.query.limit ?? '25'), 10);
    const limit = Number.isFinite(rawLimit) ? Math.max(1, Math.min(100, rawLimit)) : 25;
    const probe = probeAeonMonitor(runtime.getGnosisRoot(), runtime.effectiveAeonMonitorBinForChild());
    res.json({
      ok: true,
      code: 'RUNTIME_STATUS',
      message: 'runtime snapshot',
      details: {
        config: runtime.getConfig(),
        runtime: {
          running: runtime.isRunning(),
          lastPid: runtime.getLastPid(),
        },
        aeonMonitorProbe: probe,
        recentEvents: fanout.snapshot(limit),
      },
    });
  });

  app.get('/mesh/events', (req, res) => {
    res.writeHead(200, {
      'Content-Type': 'text/event-stream',
      'Cache-Control': 'no-cache, no-transform',
      Connection: 'keep-alive',
    });
    fanout.add(res);
    fanout.broadcast({
      type: 'mesh_events_ready',
      source: 'main',
      ts: new Date().toISOString(),
    });
    req.on('close', () => {
      fanout.remove(res);
    });
  });

  app.get('/mcp', async (_req, res) => {
    const transport = new SSEServerTransport('/messages', res);
    const sessionId = transport.sessionId;
    transports[sessionId] = transport;
    transport.onclose = () => {
      delete transports[sessionId];
    };
    const mcp = meshMcpServer(runtime, fanout);
    await mcp.connect(transport);
  });

  app.post('/messages', async (req, res) => {
    const sessionId = req.query.sessionId;
    if (typeof sessionId !== 'string' || !sessionId) {
      res.status(400).send('Missing sessionId query parameter');
      return;
    }
    const transport = transports[sessionId];
    if (!transport) {
      res.status(404).send('Unknown MCP session');
      return;
    }
    await transport.handlePostMessage(req, res, req.body);
  });

  return { app, transports };
}

/**
 * Handles the gnosis start Mesh Local Mcp workflow.
 */
export async function startMeshLocalMcp(
  opts: MeshLocalMcpListenOptions = {},
): Promise<MeshLocalMcpHandle> {
  const host = opts.host ?? process.env.MESH_LOCAL_MCP_HOST ?? '127.0.0.1';
  const port = opts.port ?? Number.parseInt(process.env.MESH_LOCAL_MCP_PORT ?? '8765', 10);
  const gnosisRoot = gnosisRootFromPackage();
  const runtime = new MeshRuntime(gnosisRoot);
  const fanout = new MeshEventFanout();
  const { app, transports } = createMeshLocalMcpApp(runtime, fanout, host);

  const server: Server = await new Promise((resolvePromise, reject) => {
    const s = app.listen(port, host, () => resolvePromise(s));
    s.on('error', reject);
  });

  const baseUrl = `http://${host === '::' ? '[::]' : host}:${port}`;
  return {
    baseUrl,
    app,
    close: async () => {
      for (const t of Object.values(transports)) {
        await t.close().catch(() => undefined);
      }
      await new Promise<void>((resolveClose, rejectClose) => {
        server.close((err) => (err ? rejectClose(err) : resolveClose()));
      });
    },
  };
}

async function main(): Promise<void> {
  const host = process.env.MESH_LOCAL_MCP_HOST ?? '127.0.0.1';
  const port = Number.parseInt(process.env.MESH_LOCAL_MCP_PORT ?? '8765', 10);
  const h = await startMeshLocalMcp({ host, port });
  // eslint-disable-next-line no-console
  console.log(
    `[gnosis-mesh-local-mcp] listening ${h.baseUrl} — MCP SSE GET /mcp, POST /messages, events GET /mesh/events`,
  );
}

const entryFile = fileURLToPath(import.meta.url);
if (process.argv[1] && resolve(process.argv[1]) === resolve(entryFile)) {
  void main().catch((e) => {
    console.error(e);
    process.exit(1);
  });
}
