// @a0n/.ignored_gnosis/mesh-local-mcp/src/mcp-register.ts
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { z } from 'zod';
import { buildAeonMonitor } from './aeon-monitor-build.js';
import { probeAeonMonitor } from './aeon-monitor-probe.js';
import type { MeshBackend, MeshProfile } from './mesh-runtime.js';
import { MeshRuntime } from './mesh-runtime.js';
import type { MeshEventFanout } from './sse-fanout.js';

const profileSchema = z.enum(['ism433', 'ism915', 'nwr162']).nullable();

const setConfigSchema = z
  .object({
    meshStationUrls: z.string().optional(),
    meshBackend: z.enum(['host', 'native', 'service', 'auto']).optional(),
    profile: profileSchema.optional(),
    gnosisAeonMonitorBin: z.string().nullable().optional(),
    monsterMeshExtraArgs: z.array(z.string()).optional(),
  })
  .strict();

const startSchema = z
  .object({
    extraArgs: z.array(z.string()).optional(),
  })
  .strict();

function textJson(value: unknown) {
  return {
    content: [{ type: 'text' as const, text: JSON.stringify(value, null, 2) }],
  };
}

function okEnvelope(
  code: string,
  message: string,
  details?: Record<string, unknown>,
): { ok: true; code: string; message: string; details?: Record<string, unknown> } {
  return { ok: true, code, message, ...(details ? { details } : {}) };
}

function errEnvelope(
  code: string,
  message: string,
  details?: Record<string, unknown>,
): { ok: false; code: string; message: string; details?: Record<string, unknown> } {
  return { ok: false, code, message, ...(details ? { details } : {}) };
}

/**
 * Handles the gnosis register Mesh Mcp Tools workflow.
 */
export function registerMeshMcpTools(
  server: McpServer,
  runtime: MeshRuntime,
  fanout: MeshEventFanout,
): void {
  server.registerTool(
    'gnosis_mesh_get_config',
    {
      description:
        'Return current mesh operator config (MESH_STATION_URLS, backend, profile, aeon-monitor path, default monster-mesh args).',
      inputSchema: z.object({}).strict(),
    },
    async () => {
      const config = runtime.getConfig();
      return textJson(okEnvelope('CONFIG_READ', 'read current operator config', { config }));
    },
  );

  server.registerTool(
    'gnosis_mesh_set_config',
    {
      description:
        'Merge live operator config used for the next gnosis_mesh_start (does not restart a running mesh).',
      inputSchema: setConfigSchema,
    },
    async (args: z.infer<typeof setConfigSchema>) => {
      const partial: {
        meshStationUrls?: string;
        meshBackend?: MeshBackend;
        profile?: MeshProfile | null;
        gnosisAeonMonitorBin?: string | null;
        monsterMeshExtraArgs?: string[];
      } = {};
      if (args.meshStationUrls !== undefined) partial.meshStationUrls = args.meshStationUrls;
      if (args.meshBackend !== undefined) partial.meshBackend = args.meshBackend;
      if (args.profile !== undefined) partial.profile = args.profile;
      if (args.gnosisAeonMonitorBin !== undefined) partial.gnosisAeonMonitorBin = args.gnosisAeonMonitorBin;
      if (args.monsterMeshExtraArgs !== undefined) {
        partial.monsterMeshExtraArgs = [...args.monsterMeshExtraArgs];
      }
      const previous = runtime.getConfig();
      const candidate = runtime.setConfig(partial);
      const invalid = runtime.validateConfig(candidate);
      if (invalid) {
        // restore previous config by re-setting the prior snapshot
        runtime.setConfig(previous);
        return textJson(errEnvelope(invalid.code, invalid.message));
      }
      fanout.broadcast({
        type: 'config_changed',
        source: 'mcp-register',
        config: candidate,
        ts: new Date().toISOString(),
      });
      return textJson(okEnvelope('CONFIG_UPDATED', 'updated operator config', { config: candidate }));
    },
  );

  server.registerTool(
    'gnosis_mesh_start',
    {
      description:
        'Spawn `bin/rtlsdr-mock-sim.js monster-mesh` with current config; stream stdout/stderr and exit on GET /mesh/events.',
      inputSchema: startSchema,
    },
    async (args: z.infer<typeof startSchema>) => {
      const r = runtime.startMesh(args.extraArgs ?? [], fanout);
      return textJson(r.ok ? okEnvelope(r.code, r.message, { pid: r.pid }) : errEnvelope(r.code, r.message));
    },
  );

  server.registerTool(
    'gnosis_mesh_stop',
    {
      description: 'SIGTERM the running monster-mesh child, if any.',
      inputSchema: z.object({}).strict(),
    },
    async () => {
      const r = runtime.stopMesh();
      fanout.broadcast({
        type: 'mesh_stop',
        source: 'mcp-register',
        ...r,
        ts: new Date().toISOString(),
      });
      return textJson(r.ok ? okEnvelope(r.code, r.message) : errEnvelope(r.code, r.message));
    },
  );

  server.registerTool(
    'gnosis_mesh_status',
    {
      description: 'Whether a mesh child is running and last known PID.',
      inputSchema: z.object({}).strict(),
    },
    async () =>
      textJson({
        ...okEnvelope('STATUS_READ', 'read mesh runtime status'),
        details: {
          running: runtime.isRunning(),
          lastPid: runtime.getLastPid(),
        },
      }),
  );

  server.registerTool(
    'gnosis_aeon_monitor_probe',
    {
      description:
        'Resolve aeon-monitor binary candidates (same rules as monster-mesh native backend): GNOSIS_AEON_MONITOR_BIN or distributed-inference/target/release, gnosis/target/release.',
      inputSchema: z.object({}).strict(),
    },
    async () => {
      const explicit = runtime.effectiveAeonMonitorBinForChild();
      const result = probeAeonMonitor(runtime.getGnosisRoot(), explicit);
      return textJson(
        result.nativeBackendReady
          ? okEnvelope('AEON_MONITOR_READY', 'aeon-monitor probe passed', { probe: result })
          : errEnvelope('AEON_MONITOR_NOT_READY', 'aeon-monitor probe failed', { probe: result }),
      );
    },
  );

  server.registerTool(
    'gnosis_aeon_monitor_set_bin',
    {
      description:
        'Set or clear GNOSIS_AEON_MONITOR_BIN override for the next mesh spawn (null clears to inherit process env only via merged spawn env).',
      inputSchema: z
        .object({
          path: z.string().nullable(),
        })
        .strict(),
    },
    async (args: { path: string | null }) => {
      const p = args.path?.trim() === '' ? null : args.path;
      const next = runtime.setConfig({ gnosisAeonMonitorBin: p });
      fanout.broadcast({
        type: 'aeon_monitor_bin_changed',
        source: 'mcp-register',
        gnosisAeonMonitorBin: next.gnosisAeonMonitorBin,
        ts: new Date().toISOString(),
      });
      return textJson(okEnvelope('AEON_MONITOR_BIN_SET', 'updated aeon-monitor override', { config: next }));
    },
  );

  server.registerTool(
    'gnosis_aeon_monitor_build',
    {
      description:
        'Run `pnpm run aeon-monitor:build` in open-source/gnosis (release aeon-monitor in distributed-inference/target). Streams no output to SSE until complete.',
      inputSchema: z.object({}).strict(),
    },
    async () => {
      fanout.broadcast({
        type: 'aeon_monitor_build_start',
        source: 'mcp-register',
        cwd: runtime.getGnosisRoot(),
        ts: new Date().toISOString(),
      });
      const r = await buildAeonMonitor(runtime.getGnosisRoot(), { timeoutMs: 120_000 });
      fanout.broadcast({
        type: 'aeon_monitor_build_finish',
        source: 'mcp-register',
        ok: r.ok,
        code: r.code,
        exitCode: r.exitCode,
        timedOut: r.timedOut,
        message: r.message,
        ts: new Date().toISOString(),
      });
      return textJson(
        r.ok
          ? okEnvelope(r.code, r.message, { exitCode: r.exitCode, timedOut: r.timedOut })
          : errEnvelope(r.code, r.message, {
              exitCode: r.exitCode,
              timedOut: r.timedOut,
              stdoutTail: r.stdout.slice(-4000),
              stderrTail: r.stderr.slice(-4000),
            }),
      );
    },
  );

  server.registerTool(
    'gnosis_mesh_runtime_snapshot',
    {
      description:
        'Snapshot runtime + config + aeon-monitor probe + recent SSE events (bounded).',
      inputSchema: z
        .object({
          recentEventLimit: z.number().int().min(1).max(100).optional(),
        })
        .strict(),
    },
    async (args: { recentEventLimit?: number }) => {
      const probe = probeAeonMonitor(
        runtime.getGnosisRoot(),
        runtime.effectiveAeonMonitorBinForChild(),
      );
      return textJson(
        okEnvelope('RUNTIME_SNAPSHOT', 'captured runtime snapshot', {
          config: runtime.getConfig(),
          runtime: {
            running: runtime.isRunning(),
            lastPid: runtime.getLastPid(),
            gnosisRoot: runtime.getGnosisRoot(),
          },
          aeonMonitorProbe: probe,
          recentEvents: fanout.snapshot(args.recentEventLimit ?? 25),
        }),
      );
    },
  );

  // ── R1 inference-fleet control ─────────────────────────────────────────
  // Full mesh control for agents, proxied to a running `aeon-monitor --mesh
  // --serve` (AEON_MONITOR_URL, default 127.0.0.1:8722) over /monitor/control —
  // the SAME surface the skymesh app uses. Ops are async; poll fleet_status.
  const monitorBase = (process.env.AEON_MONITOR_URL || 'http://127.0.0.1:8722').replace(/\/$/, '');
  const monitorControl = async (op: string, extra: Record<string, unknown> = {}): Promise<unknown> => {
    const r = await fetch(`${monitorBase}/monitor/control`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ op, ...extra }),
    });
    return (await r.json()) as unknown;
  };

  server.registerTool(
    'gnosis_mesh_fleet_status',
    {
      description:
        'Live R1 inference-fleet status from aeon-monitor --mesh (nodes, control surface, last action). Needs `aeon-monitor --mesh --serve` (AEON_MONITOR_URL, default 127.0.0.1:8722).',
      inputSchema: z.object({}).strict(),
    },
    async () => {
      try {
        const r = await fetch(`${monitorBase}/monitor/status`);
        const s = (await r.json()) as { mesh?: unknown };
        return textJson(okEnvelope('FLEET_STATUS', 'read fleet status', { mesh: s.mesh ?? null }));
      } catch (e) {
        return textJson(
          errEnvelope('FLEET_UNREACHABLE', `aeon-monitor not reachable at ${monitorBase}`, {
            error: String(e),
          }),
        );
      }
    },
  );

  const nodeSchema = z.object({ node: z.string() }).strict();
  const nodeOps: ReadonlyArray<readonly [string, string, string]> = [
    ['gnosis_mesh_node_probe', 'mesh_probe', 'probe a node (/chain -> latency + argmax)'],
    ['gnosis_mesh_node_restart', 'mesh_restart', "restart a node's in-app launcher (adb)"],
    ['gnosis_mesh_node_reboot', 'mesh_reboot', 'reboot a node device (adb)'],
  ];
  for (const [tool, op, verb] of nodeOps) {
    server.registerTool(
      tool,
      {
        description: `Fleet control: ${verb}. Async — poll gnosis_mesh_fleet_status (mesh.lastAction) for the result.`,
        inputSchema: nodeSchema,
      },
      async (args: z.infer<typeof nodeSchema>) => {
        try {
          const res = await monitorControl(op, { node: args.node });
          return textJson(okEnvelope('FLEET_OP', `dispatched ${op} on ${args.node}`, { res }));
        } catch (e) {
          return textJson(
            errEnvelope('FLEET_UNREACHABLE', `aeon-monitor not reachable at ${monitorBase}`, {
              error: String(e),
            }),
          );
        }
      },
    );
  }

  server.registerTool(
    'gnosis_mesh_fleet_bench',
    {
      description:
        'Bench the whole fleet (sequential vs concurrent /chain -> throughput + speedup). Async — poll gnosis_mesh_fleet_status (mesh.lastAction).',
      inputSchema: z.object({}).strict(),
    },
    async () => {
      try {
        const res = await monitorControl('mesh_bench');
        return textJson(okEnvelope('FLEET_BENCH', 'dispatched fleet bench', { res }));
      } catch (e) {
        return textJson(
          errEnvelope('FLEET_UNREACHABLE', `aeon-monitor not reachable at ${monitorBase}`, {
            error: String(e),
          }),
        );
      }
    },
  );
}
