// @a0n/.ignored_gnosis/mesh-local-mcp/src/mesh-runtime.ts
import { spawn, type ChildProcess } from 'node:child_process';
import { resolve } from 'node:path';
import { findExecutableAeonMonitor } from './aeon-monitor-probe.js';
import type { MeshEventFanout } from './sse-fanout.js';

export type MeshProfile = 'ism433' | 'ism915' | 'nwr162';
export type MeshBackend = 'host' | 'native' | 'service' | 'auto';

export interface MeshOperatorConfig {
  readonly meshStationUrls: string;
  readonly meshBackend: MeshBackend;
  readonly profile: MeshProfile | null;
  readonly gnosisAeonMonitorBin: string | null;
  readonly monsterMeshExtraArgs: readonly string[];
}

export interface MeshActionResult {
  readonly ok: boolean;
  readonly code: string;
  readonly message: string;
  readonly details?: Record<string, unknown>;
  readonly pid?: number;
}

const defaultConfig = (): MeshOperatorConfig => ({
  meshStationUrls: process.env.MESH_STATION_URLS ?? '',
  meshBackend: (process.env.GNOSIS_MONSTER_MESH_BACKEND as MeshBackend | undefined) ?? 'auto',
  profile: null,
  gnosisAeonMonitorBin: process.env.GNOSIS_AEON_MONITOR_BIN?.trim() || null,
  monsterMeshExtraArgs: [],
});

function profileEnv(profile: MeshProfile | null): Record<string, string> {
  if (!profile) return {};
  switch (profile) {
    case 'ism433':
      return { GNOSIS_MESH_PROFILE: 'ism433', GNOSIS_MESH_CENTER_MHZ: '433.92' };
    case 'ism915':
      return { GNOSIS_MESH_PROFILE: 'ism915', GNOSIS_MESH_CENTER_MHZ: '915' };
    case 'nwr162':
      return { GNOSIS_MESH_PROFILE: 'nwr162', GNOSIS_MESH_CENTER_MHZ: '162.475' };
    default:
      return {};
  }
}

export class MeshRuntime {
  private config: MeshOperatorConfig;
  private child: ChildProcess | null = null;
  private lastPid: number | null = null;
  private readonly gnosisRoot: string;
  private readonly rtlSimJs: string;

  constructor(gnosisRoot: string) {
    this.gnosisRoot = gnosisRoot;
    this.rtlSimJs = resolve(gnosisRoot, 'bin/rtlsdr-mock-sim.js');
    this.config = defaultConfig();
  }

  getConfig(): MeshOperatorConfig {
    return { ...this.config, monsterMeshExtraArgs: [...this.config.monsterMeshExtraArgs] };
  }

  setConfig(partial: Partial<MeshOperatorConfig>): MeshOperatorConfig {
    this.config = {
      meshStationUrls: partial.meshStationUrls ?? this.config.meshStationUrls,
      meshBackend: partial.meshBackend ?? this.config.meshBackend,
      profile: partial.profile !== undefined ? partial.profile : this.config.profile,
      gnosisAeonMonitorBin:
        partial.gnosisAeonMonitorBin !== undefined
          ? partial.gnosisAeonMonitorBin
          : this.config.gnosisAeonMonitorBin,
      monsterMeshExtraArgs:
        partial.monsterMeshExtraArgs !== undefined
          ? [...partial.monsterMeshExtraArgs]
          : [...this.config.monsterMeshExtraArgs],
    };
    return this.getConfig();
  }

  isRunning(): boolean {
    return this.child !== null && !this.child.killed;
  }

  getLastPid(): number | null {
    return this.lastPid;
  }

  getGnosisRoot(): string {
    return this.gnosisRoot;
  }

  /**
   * Value that will apply to the next `monster-mesh` child as `GNOSIS_AEON_MONITOR_BIN`
   * when only one path should be probed (matches `wrapper-mesh-delegation.mjs`).
   */
  effectiveAeonMonitorBinForChild(): string | null {
    const fromConfig = this.config.gnosisAeonMonitorBin?.trim();
    if (fromConfig) return fromConfig;
    return process.env.GNOSIS_AEON_MONITOR_BIN?.trim() ?? null;
  }

  validateConfig(next: MeshOperatorConfig): MeshActionResult | null {
    if ((next.meshBackend === 'host' || next.meshBackend === 'service') && !next.meshStationUrls.trim()) {
      return {
        ok: false,
        code: 'MESH_STATION_URLS_REQUIRED',
        message: `${next.meshBackend} backend requires non-empty meshStationUrls`,
      };
    }

    if (next.meshBackend === 'native') {
      const explicit = next.gnosisAeonMonitorBin?.trim() || this.effectiveAeonMonitorBinForChild();
      const resolved = findExecutableAeonMonitor(this.gnosisRoot, explicit);
      if (!resolved) {
        return {
          ok: false,
          code: 'AEON_MONITOR_NOT_EXECUTABLE',
          message:
            'native backend requires executable aeon-monitor (set gnosisAeonMonitorBin/GNOSIS_AEON_MONITOR_BIN or build aeon-monitor)',
        };
      }
    }

    return null;
  }

  stopMesh(): MeshActionResult {
    if (!this.child || this.child.killed) {
      return {
        ok: false,
        code: 'MESH_NOT_RUNNING',
        message: 'no mesh child process',
      };
    }
    this.child.kill('SIGTERM');
    this.child = null;
    this.lastPid = null;
    return {
      ok: true,
      code: 'STOP_SIGNAL_SENT',
      message: 'sent SIGTERM to mesh child',
    };
  }

  startMesh(
    extraArgs: readonly string[],
    fanout: MeshEventFanout,
  ): MeshActionResult {
    if (this.isRunning()) {
      return {
        ok: false,
        code: 'MESH_ALREADY_RUNNING',
        message: 'mesh already running; call gnosis_mesh_stop first',
      };
    }

    const invalidConfig = this.validateConfig(this.config);
    if (invalidConfig) {
      return invalidConfig;
    }

    const argv = [this.rtlSimJs, 'monster-mesh', '--mesh-backend', this.config.meshBackend];
    if (this.config.profile) {
      argv.push('--profile', this.config.profile);
    }
    argv.push(...this.config.monsterMeshExtraArgs, ...extraArgs);

    const env: NodeJS.ProcessEnv = {
      ...process.env,
      MESH_STATION_URLS: this.config.meshStationUrls,
      ...profileEnv(this.config.profile),
    };
    if (this.config.meshBackend !== 'auto') {
      env.GNOSIS_MONSTER_MESH_BACKEND = this.config.meshBackend;
    } else {
      delete env.GNOSIS_MONSTER_MESH_BACKEND;
    }
    if (this.config.gnosisAeonMonitorBin) {
      env.GNOSIS_AEON_MONITOR_BIN = this.config.gnosisAeonMonitorBin;
    }

    fanout.broadcast({
      type: 'mesh_spawn',
      argv,
      cwd: this.gnosisRoot,
      ts: new Date().toISOString(),
    });

    const child = spawn(process.execPath, argv, {
      cwd: this.gnosisRoot,
      env,
      stdio: ['ignore', 'pipe', 'pipe'],
    });
    this.child = child;
    this.lastPid = child.pid ?? null;

    const onChunk = (stream: 'stdout' | 'stderr') => (chunk: Buffer) => {
      const text = chunk.toString('utf8');
      fanout.broadcast({ type: `mesh_${stream}`, text, ts: new Date().toISOString() });
    };
    child.stdout?.on('data', onChunk('stdout'));
    child.stderr?.on('data', onChunk('stderr'));

    child.on('exit', (code, signal) => {
      this.child = null;
      this.lastPid = null;
      fanout.broadcast({
        type: 'mesh_exit',
        source: 'mesh-runtime',
        code,
        signal,
        ts: new Date().toISOString(),
      });
    });

    child.on('error', (err) => {
      this.child = null;
      this.lastPid = null;
      fanout.broadcast({
        type: 'mesh_error',
        source: 'mesh-runtime',
        message: err instanceof Error ? err.message : String(err),
        ts: new Date().toISOString(),
      });
    });

    return {
      ok: true,
      code: 'MESH_STARTED',
      pid: child.pid,
      message: 'spawned monster-mesh child',
    };
  }
}
