// @a0n/.ignored_gnosis/mesh-local-mcp/src/sse-fanout.ts
import type { Response } from 'express';

/** Minimal fanout for `text/event-stream` (custom mesh telemetry, not MCP wire). */
export class MeshEventFanout {
  private readonly clients = new Set<Response>();
  private readonly recentEvents: Record<string, unknown>[] = [];
  private readonly maxRecent: number;

  constructor(maxRecent: number = 100) {
    this.maxRecent = maxRecent;
  }

  add(res: Response): void {
    this.clients.add(res);
  }

  remove(res: Response): void {
    this.clients.delete(res);
  }

  broadcast(event: Record<string, unknown>): void {
    const normalized = {
      eventType: String(event.type ?? event.eventType ?? 'mesh_event'),
      source: String(event.source ?? 'mesh-local-mcp'),
      ts: String(event.ts ?? new Date().toISOString()),
      ...event,
    };
    this.recentEvents.push(normalized);
    if (this.recentEvents.length > this.maxRecent) {
      this.recentEvents.splice(0, this.recentEvents.length - this.maxRecent);
    }

    const payload = `data: ${JSON.stringify(normalized)}\n\n`;
    for (const res of this.clients) {
      try {
        res.write(payload);
      } catch {
        this.clients.delete(res);
      }
    }
  }

  snapshot(limit: number = 25): readonly Record<string, unknown>[] {
    const n = Math.max(1, Math.min(this.maxRecent, Math.floor(limit)));
    return this.recentEvents.slice(-n);
  }
}
