// @a0n/.ignored_gnosis/forkjoin-app/client/components/Sections/AeonSection.tsx
/**
 * AeonSection - Distributed Sync Standard
 *
 * Shows the incredible capabilities from the Aeon README:
 * - Distributed Synchronization (SyncCoordinator, Heartbeat)
 * - Schema Versioning (Migrations, Rollbacks, Compatibility)
 * - Conflict Resolution (LWW, Vector Clock, Majority Vote)
 * - Data Replication (Policies, Consistency Levels)
 * - Offline-First (Operation Queuing, Encryption)
 * - Compression (Delta Sync, 70-90% reduction)
 */

import {
  RefreshCw,
  GitBranch,
  Swords,
  Users,
  WifiOff,
  PackageOpen,
  Network,
  Clock,
  Shield,
  Database,
  Zap,
  Heart,
  History,
  Lock,
  Layers,
  Gauge,
} from '@a0n/aeon-icons';

const features = [
  // Row 1: Core Sync
  {
    icon: RefreshCw,
    title: 'Sync Coordinator',
    metric: 'Multi-node',
    description:
      'Create and manage sync sessions across multiple nodes. No central server required.',
    category: 'core',
  },
  {
    icon: Heart,
    title: 'Heartbeat Monitoring',
    metric: 'Real-time',
    description:
      'Automatic health monitoring with configurable intervals. Detect failures instantly.',
    category: 'core',
  },
  {
    icon: Network,
    title: 'Sync Protocol',
    metric: 'Handshake',
    description:
      'Protocol messages for session establishment. Full handshaking and state exchange.',
    category: 'core',
  },
  {
    icon: Clock,
    title: 'Event-Driven',
    metric: 'EventEmitter',
    description:
      'Subscribe to sync-completed, conflicts, node status. Reactive architecture.',
    category: 'core',
  },

  // Row 2: Versioning & Migrations
  {
    icon: GitBranch,
    title: 'Schema Versioning',
    metric: 'SemVer',
    description:
      'Track schema versions with major.minor.patch. Breaking change detection built in.',
    category: 'versioning',
  },
  {
    icon: History,
    title: 'Migration Engine',
    metric: 'Up/Down',
    description:
      'Execute migrations forward or rollback. Full audit trail of all changes.',
    category: 'versioning',
  },
  {
    icon: Layers,
    title: 'Data Transformer',
    metric: 'Schema',
    description:
      'Transform data during migrations. Map old schemas to new seamlessly.',
    category: 'versioning',
  },
  {
    icon: Shield,
    title: 'Compatibility Check',
    metric: 'Auto',
    description:
      'Check migration paths between versions. Prevent incompatible upgrades.',
    category: 'versioning',
  },

  // Row 3: Conflict Resolution
  {
    icon: Swords,
    title: 'Last-Write-Wins',
    metric: 'LWW',
    description:
      'Simple timestamp-based resolution. Fast and predictable for most cases.',
    category: 'resolution',
  },
  {
    icon: Clock,
    title: 'Vector Clocks',
    metric: 'Causal',
    description:
      'Causal ordering across nodes. Detect concurrent edits with precision.',
    category: 'resolution',
  },
  {
    icon: Users,
    title: 'Majority Vote',
    metric: 'Quorum',
    description:
      'Consensus-based resolution. Most common value wins across replicas.',
    category: 'resolution',
  },
  {
    icon: Database,
    title: 'State Reconciler',
    metric: 'Diff',
    description:
      'Generate diffs between divergent states. Merge with configurable strategies.',
    category: 'resolution',
  },

  // Row 4: Optimization & Offline
  {
    icon: WifiOff,
    title: 'Offline Queue',
    metric: 'Priority',
    description:
      'Priority-based operation queue with retry logic. AES-256-GCM encryption at rest.',
    category: 'optimization',
  },
  {
    icon: PackageOpen,
    title: 'Delta Compression',
    metric: '70-90%',
    description:
      'Field-level change detection. Send only what changed, not the whole document.',
    category: 'optimization',
  },
  {
    icon: Zap,
    title: 'Prefetching Engine',
    metric: 'Predictive',
    description:
      'Pre-compress based on operation patterns. Anticipate what you will sync next.',
    category: 'optimization',
  },
  {
    icon: Gauge,
    title: 'Batch Optimizer',
    metric: 'Adaptive',
    description:
      'Intelligent batch scheduling based on network conditions. Auto-adjusting.',
    category: 'optimization',
  },
];

const categoryColors: Record<string, string> = {
  core: '#5599cc',
  versioning: '#10b981',
  resolution: '#f59e0b',
  optimization: '#06b6d4',
};

/**
 * Renders the Aeon Section view.
 */
export function AeonSection() {
  return (
    <section
      id="aeon"
      className="section aeon-section"
      aria-labelledby="aeon-title"
    >
      <div className="container">
        <div className="section-header">
          <span className="layer-label">Sync Layer</span>
          <h2 id="aeon-title" className="product-section-title">
            Aeon
          </h2>
          <p className="section-description">
            Distributed synchronization, schema versioning, and conflict
            resolution.
          </p>
        </div>

        <div className="aeon-grid">
          {features.map((feature) => (
            <div
              key={feature.title}
              className="aeon-tile"
              style={
                {
                  '--tile-color': categoryColors[feature.category],
                } as React.CSSProperties
              }
            >
              <feature.icon size={24} className="tile-icon" />
              <div className="tile-content">
                <div className="tile-header">
                  <h3>{feature.title}</h3>
                  <span className="tile-metric">{feature.metric}</span>
                </div>
                <p>{feature.description}</p>
              </div>
            </div>
          ))}
        </div>

        <div className="comparison-table">
          <h3>vs Other Libraries</h3>
          <table className="comparison-grid-table">
            <thead>
              <tr>
                <th>Feature</th>
                <th>Aeon</th>
                <th>Yjs</th>
                <th>Automerge</th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>Schema Versioning</td>
                <td className="yes">✓ Yes</td>
                <td className="no">No</td>
                <td className="no">No</td>
              </tr>
              <tr>
                <td>Migrations</td>
                <td className="yes">✓ Yes</td>
                <td className="no">No</td>
                <td className="no">No</td>
              </tr>
              <tr>
                <td>Replication Policies</td>
                <td className="yes">✓ Yes</td>
                <td className="no">No</td>
                <td className="no">No</td>
              </tr>
              <tr>
                <td>Multiple Merge Strategies</td>
                <td className="yes">✓ Yes</td>
                <td className="partial">Partial</td>
                <td className="partial">Partial</td>
              </tr>
              <tr>
                <td>Zero Dependencies</td>
                <td className="yes">✓ Yes</td>
                <td className="no">No</td>
                <td className="no">No</td>
              </tr>
            </tbody>
          </table>
        </div>

        <div className="cta-row">
          <a href="#/aeon" className="button button-primary">
            Explore Aeon
          </a>
        </div>
      </div>

      <style>{`
        .aeon-section {
          background: var(--color-bg);
          border-bottom: 1px solid var(--color-bg-tertiary);
        }

        .section-header {
          text-align: center;
          margin-bottom: var(--space-3xl);
        }

        .section-description {
          font-size: 1.125rem;
          color: var(--color-text-secondary);
          max-width: 600px;
          margin: var(--space-md) auto 0;
          line-height: 1.6;
        }

        .aeon-grid {
          display: grid;
          grid-template-columns: repeat(4, 1fr);
          gap: var(--space-md);
          margin-bottom: var(--space-3xl);
        }

        .aeon-tile {
          display: flex;
          gap: var(--space-md);
          padding: var(--space-md);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 10px;
          transition: all 0.2s ease;
        }

        .aeon-tile:hover {
          border-color: var(--tile-color);
          transform: translateY(-2px);
        }

        .tile-icon {
          color: var(--tile-color);
          flex-shrink: 0;
          margin-top: 2px;
        }

        .tile-content {
          flex: 1;
          min-width: 0;
        }

        .tile-header {
          display: flex;
          align-items: center;
          justify-content: space-between;
          gap: var(--space-sm);
          margin-bottom: var(--space-xs);
        }

        .tile-header h3 {
          font-size: 0.9rem;
          font-weight: 600;
          color: var(--color-text);
          margin: 0;
          white-space: nowrap;
          overflow: hidden;
          text-overflow: ellipsis;
        }

        .tile-metric {
          font-family: var(--font-mono);
          font-size: 0.7rem;
          font-weight: 600;
          color: var(--tile-color);
          padding: 2px 6px;
          background: rgba(255, 255, 255, 0.05);
          border-radius: 4px;
          flex-shrink: 0;
        }

        .aeon-tile p {
          font-size: 0.8rem;
          color: var(--color-text-muted);
          margin: 0;
          line-height: 1.4;
        }

        .comparison-table {
          max-width: 700px;
          margin: 0 auto var(--space-3xl);
          text-align: center;
        }

        .comparison-table h3 {
          margin-bottom: var(--space-lg);
          color: var(--color-text-muted);
          text-transform: uppercase;
          letter-spacing: 0.1em;
          font-size: 0.85rem;
        }

        .comparison-grid-table {
          width: 100%;
          border-collapse: collapse;
          border-radius: 8px;
          overflow: hidden;
          background: var(--color-bg-tertiary);
        }

        .comparison-grid-table th,
        .comparison-grid-table td {
          padding: var(--space-sm) var(--space-md);
          font-size: 0.85rem;
          border: 1px solid var(--color-bg-tertiary);
        }

        .comparison-grid-table th {
          background: var(--color-bg-secondary);
          font-weight: 600;
          text-align: center;
        }

        .comparison-grid-table td {
          background: var(--color-bg);
        }

        .comparison-grid-table td:first-child {
          text-align: left;
          color: var(--color-text-secondary);
        }

        .comparison-grid-table td:not(:first-child) {
          text-align: center;
          font-weight: 600;
        }

        .comparison-grid-table .yes {
          color: #10b981;
        }

        .comparison-grid-table .no {
          color: var(--color-text-muted);
        }

        .comparison-grid-table .partial {
          color: #f59e0b;
        }

        .cta-row {
          display: flex;
          justify-content: center;
        }

        @media (max-width: 1200px) {
          .aeon-grid {
            grid-template-columns: repeat(2, 1fr);
          }
        }

        @media (max-width: 600px) {
          .aeon-grid {
            grid-template-columns: 1fr;
          }

          .aeon-tile {
            flex-direction: column;
            text-align: center;
          }

          .tile-header {
            justify-content: center;
          }

          .tile-icon {
            margin: 0 auto;
          }

          .comparison-grid-table th,
          .comparison-grid-table td {
            padding: var(--space-xs) var(--space-sm);
            font-size: 0.75rem;
          }
        }
      `}</style>
    </section>
  );
}
