// @a0n/.ignored_gnosis/forkjoin-app/client/components/Sections/AuthSection.tsx
/**
 * AuthSection - Aegis (UCAN + ZK Auth)
 *
 * Decentralized identity and fine-grained access control:
 * - UCAN Tokens (User Controlled Authorization Networks)
 * - XPath-like Node Selection for surgical access control
 * - Firebase-Style Rules with expressions
 * - Zero-Knowledge Proofs (prove without revealing)
 * - DID:key Identity (Ed25519/P-256)
 */

import {
  Lock,
  Eye,
  Key,
  Shield,
  UserCheck,
  FileCode,
  GitBranch,
  Timer,
  Fingerprint,
  Database,
  Settings,
  Layers,
  Share2,
  CheckCircle,
  XCircle,
  Code,
} from '@a0n/aeon-icons';

const features = [
  // Row 1: Identity
  {
    icon: Fingerprint,
    title: 'DID:key Identity',
    metric: 'Ed25519',
    description:
      'Generate identities with Web Crypto API. No server, no signup, no tracking.',
    category: 'identity',
  },
  {
    icon: Key,
    title: 'Keypair Generation',
    metric: 'Client-side',
    description:
      'Ed25519 or P-256 curves. Signing and optional encryption keys included.',
    category: 'identity',
  },
  {
    icon: Lock,
    title: 'Sign & Verify',
    metric: 'Crypto',
    description:
      'Sign any data with your identity. Verify signatures from any source.',
    category: 'identity',
  },
  {
    icon: Database,
    title: 'Secure Storage',
    metric: 'IndexedDB',
    description:
      'Persist identities in IndexedDB or memory. Your keys never leave the device.',
    category: 'identity',
  },

  // Row 2: UCAN
  {
    icon: Shield,
    title: 'UCAN Tokens',
    metric: 'JWT',
    description:
      'Create, parse, verify capability tokens. Offline-first verification.',
    category: 'ucan',
  },
  {
    icon: Share2,
    title: 'Delegation',
    metric: 'Attenuate',
    description:
      'Delegate capabilities with attenuation. Pass only what you have.',
    category: 'ucan',
  },
  {
    icon: Timer,
    title: 'Expiring Grants',
    metric: 'TTL',
    description:
      'Time-limited capabilities. Auto-expire after seconds, hours, or days.',
    category: 'ucan',
  },
  {
    icon: Layers,
    title: 'Capability Chain',
    metric: 'Proofs',
    description:
      'Chain of authority from root to leaf. Verify the full delegation path.',
    category: 'ucan',
  },

  // Row 3: Access Control
  {
    icon: Code,
    title: 'XPath Selection',
    metric: '//path/*',
    description:
      'Target specific nodes with XPath-like expressions. Surgical access control.',
    category: 'access',
  },
  {
    icon: CheckCircle,
    title: 'Grant Access',
    metric: 'Allow',
    description:
      'Grant read, write, delete, admin per path. Wildcards supported.',
    category: 'access',
  },
  {
    icon: XCircle,
    title: 'Deny Access',
    metric: 'Block',
    description:
      'Deny with priority. Higher priority rules win. Default-deny patterns.',
    category: 'access',
  },
  {
    icon: Settings,
    title: 'Predicates',
    metric: '[filter]',
    description:
      'Filter by property: [status="published"], [author=$userId], [price>100].',
    category: 'access',
  },

  // Row 4: Advanced
  {
    icon: FileCode,
    title: 'Firebase Rules',
    metric: '.read/.write',
    description:
      'Declarative security rules. Expression syntax for complex conditions.',
    category: 'advanced',
  },
  {
    icon: UserCheck,
    title: 'Sandboxing',
    metric: 'Zones',
    description:
      'Create public sandboxes, user-owned areas, read-only zones with one call.',
    category: 'advanced',
  },
  {
    icon: Eye,
    title: 'Zero-Knowledge',
    metric: 'ZK',
    description:
      'Prove without revealing. Age, membership, ownership without exposing data.',
    category: 'advanced',
  },
  {
    icon: GitBranch,
    title: 'Export/Import',
    metric: 'JSON',
    description:
      'Export rules to JSON. Import to other instances. Version control friendly.',
    category: 'advanced',
  },
];

const categoryColors: Record<string, string> = {
  identity: '#ec4899',
  ucan: '#10b981',
  access: '#f59e0b',
  advanced: '#5599cc',
};

/**
 * Renders the Auth Section view.
 */
export function AuthSection() {
  return (
    <section
      id="aegis"
      className="section auth-section"
      aria-labelledby="aegis-title"
    >
      <div className="container">
        <div className="section-header">
          <span className="layer-label">Identity Layer</span>
          <h2 id="aegis-title" className="product-section-title">
            Aegis
          </h2>
          <p className="section-description">
            Decentralized auth. UCAN capabilities. Zero-knowledge proofs.
            Vault-backed trust primitives.
          </p>
        </div>

        <div className="auth-grid">
          {features.map((feature) => (
            <div
              key={feature.title}
              className="auth-tile"
              style={
                {
                  '--tile-color': categoryColors[feature.category],
                } as React.CSSProperties
              }
            >
              <feature.icon size={24} className="tile-icon" />
              <div className="tile-content">
                <div className="tile-header">
                  <h3>{feature.title}</h3>
                  <span className="tile-metric">{feature.metric}</span>
                </div>
                <p>{feature.description}</p>
              </div>
            </div>
          ))}
        </div>

        <div className="comparison-table">
          <h3>vs Other Auth Providers</h3>
          <table className="comparison-grid-table">
            <thead>
              <tr>
                <th>Feature</th>
                <th>Aegis</th>
                <th>Firebase</th>
                <th>Auth0</th>
                <th>Clerk</th>
                <th>Magic.link</th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>Decentralized</td>
                <td className="yes">✓ No server</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="partial">Partial</td>
              </tr>
              <tr>
                <td>UCAN Tokens</td>
                <td className="yes">✓ Native</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
              </tr>
              <tr>
                <td>DID:key Identity</td>
                <td className="yes">✓ Yes</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="partial">DID support</td>
              </tr>
              <tr>
                <td>Zero-Knowledge Proofs</td>
                <td className="yes">✓ Built-in</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
              </tr>
              <tr>
                <td>Capability Delegation</td>
                <td className="yes">✓ Attenuate</td>
                <td className="no">No</td>
                <td className="partial">Basic</td>
                <td className="partial">Basic</td>
                <td className="no">No</td>
              </tr>
              <tr>
                <td>Offline Verification</td>
                <td className="yes">✓ Yes</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
              </tr>
              <tr>
                <td>XPath-like Rules</td>
                <td className="yes">✓ Yes</td>
                <td className="partial">Basic rules</td>
                <td className="partial">RBAC</td>
                <td className="partial">RBAC</td>
                <td className="no">No</td>
              </tr>
              <tr>
                <td>Self-Hosted</td>
                <td className="yes">✓ Yes</td>
                <td className="no">No</td>
                <td className="partial">Enterprise</td>
                <td className="no">No</td>
                <td className="no">No</td>
              </tr>
            </tbody>
          </table>
        </div>

        <div className="code-example">
          <h3>XPath-like Access Control</h3>
          <pre>
            <code>{`// Authors can edit their own posts
ac.grant(authorDID, '//posts/*[author=$authorId]/**', ['read', 'write']);

// Editors can edit content but not metadata
ac.grant(editorDID, '//posts/*/content', ['read', 'write']);

// Public can read published posts
ac.grantPublic('//posts/*[status="published"]', 'read');`}</code>
          </pre>
        </div>

        <div className="cta-row">
          <a href="#/aegis" className="button button-primary">
            Explore Aegis
          </a>
        </div>
      </div>

      <style>{`
        .auth-section {
          background: linear-gradient(180deg, var(--color-bg) 0%, var(--color-bg-secondary) 100%);
        }

        .section-header {
          text-align: center;
          margin-bottom: var(--space-3xl);
        }

        .section-description {
          font-size: 1.125rem;
          color: var(--color-text-secondary);
          max-width: 600px;
          margin: var(--space-md) auto 0;
        }

        .auth-grid {
          display: grid;
          grid-template-columns: repeat(4, 1fr);
          gap: var(--space-md);
          margin-bottom: var(--space-3xl);
        }

        .auth-tile {
          display: flex;
          gap: var(--space-md);
          padding: var(--space-md);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 10px;
          transition: all 0.2s ease;
        }

        .auth-tile:hover {
          border-color: var(--tile-color);
          transform: translateY(-2px);
        }

        .tile-icon {
          color: var(--tile-color);
          flex-shrink: 0;
          margin-top: 2px;
        }

        .tile-content {
          flex: 1;
          min-width: 0;
        }

        .tile-header {
          display: flex;
          align-items: center;
          justify-content: space-between;
          gap: var(--space-sm);
          margin-bottom: var(--space-xs);
        }

        .tile-header h3 {
          font-size: 0.9rem;
          font-weight: 600;
          color: var(--color-text);
          margin: 0;
          white-space: nowrap;
          overflow: hidden;
          text-overflow: ellipsis;
        }

        .tile-metric {
          font-family: var(--font-mono);
          font-size: 0.7rem;
          font-weight: 600;
          color: var(--tile-color);
          padding: 2px 6px;
          background: rgba(255, 255, 255, 0.05);
          border-radius: 4px;
          flex-shrink: 0;
        }

        .auth-tile p {
          font-size: 0.8rem;
          color: var(--color-text-muted);
          margin: 0;
          line-height: 1.4;
        }

        .comparison-table {
          max-width: 900px;
          margin: 0 auto var(--space-3xl);
          text-align: center;
        }

        .comparison-table h3 {
          margin-bottom: var(--space-lg);
          color: var(--color-text-muted);
          text-transform: uppercase;
          letter-spacing: 0.1em;
          font-size: 0.85rem;
        }

        .comparison-grid-table {
          width: 100%;
          border-collapse: collapse;
          border-radius: 8px;
          overflow: hidden;
          background: var(--color-bg-tertiary);
        }

        .comparison-grid-table th,
        .comparison-grid-table td {
          padding: var(--space-sm) var(--space-md);
          font-size: 0.8rem;
          border: 1px solid var(--color-bg-tertiary);
        }

        .comparison-grid-table th {
          background: var(--color-bg-secondary);
          font-weight: 600;
          text-align: center;
        }

        .comparison-grid-table td {
          background: var(--color-bg);
        }

        .comparison-grid-table td:first-child {
          text-align: left;
          color: var(--color-text-secondary);
        }

        .comparison-grid-table td:not(:first-child) {
          text-align: center;
          font-weight: 600;
        }

        .comparison-grid-table .yes {
          color: #10b981;
        }

        .comparison-grid-table .no {
          color: var(--color-text-muted);
        }

        .comparison-grid-table .partial {
          color: #f59e0b;
        }

        .code-example {
          max-width: 700px;
          margin: 0 auto var(--space-3xl);
          text-align: center;
        }

        .code-example h3 {
          margin-bottom: var(--space-md);
          color: var(--color-text-muted);
          text-transform: uppercase;
          letter-spacing: 0.1em;
          font-size: 0.85rem;
        }

        .code-example pre {
          background: var(--color-bg);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 8px;
          padding: var(--space-lg);
          text-align: left;
          overflow-x: auto;
        }

        .code-example code {
          font-family: var(--font-mono);
          font-size: 0.85rem;
          color: var(--color-text-secondary);
          line-height: 1.6;
        }

        .cta-row {
          display: flex;
          justify-content: center;
        }

        @media (max-width: 1200px) {
          .auth-grid {
            grid-template-columns: repeat(2, 1fr);
          }
        }

        @media (max-width: 600px) {
          .auth-grid {
            grid-template-columns: 1fr;
          }

          .auth-tile {
            flex-direction: column;
            text-align: center;
          }

          .tile-header {
            justify-content: center;
          }

          .tile-icon {
            margin: 0 auto;
          }

          .comparison-table {
            overflow-x: auto;
          }

          .comparison-grid-table th,
          .comparison-grid-table td {
            padding: var(--space-xs) var(--space-sm);
            font-size: 0.7rem;
          }

          .code-example pre {
            font-size: 0.75rem;
            padding: var(--space-md);
          }
        }
      `}</style>
    </section>
  );
}
