// @a0n/.ignored_gnosis/forkjoin-app/client/components/Sections/DemoSection.tsx
import { memo, useState } from '@a0n/raect';
import { Globe } from '@a0n/aeon-icons';

export const DemoSection = memo(function DemoSection() {
  return (
    <section id="demo" className="section demo-section">
      <div className="container">
        <div className="section-header">
          <h2>See It Work</h2>
          <p className="section-subtitle">
            This page is built with Aeon and Aeon-Flux. Every cursor you see is
            real. Every interaction is synced.
          </p>
        </div>

        <div className="demo-grid">
          <CRDTDemo />
          <PresenceDemo />
          <OfflineDemo />
        </div>
      </div>

      <style>{`
        .demo-section {
          background: var(--color-bg);
          border-bottom: 1px solid var(--color-bg-tertiary);
        }

        .section-header {
          text-align: center;
          margin-bottom: var(--space-3xl);
        }

        .section-badge {
          display: inline-block;
          padding: 0.25rem 0.75rem;
          font-size: 0.75rem;
          font-weight: 500;
          text-transform: uppercase;
          letter-spacing: 0.1em;
          color: var(--color-success);
          border: 1px solid var(--color-success);
          border-radius: var(--radius-full);
          margin-bottom: var(--space-md);
        }

        .section-subtitle {
          font-size: 1.25rem;
          max-width: 600px;
          margin: 0 auto;
        }

        .demo-grid {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(300px, 1fr));
          gap: var(--space-xl);
        }
      `}</style>
    </section>
  );
});

const CRDTDemo = memo(function CRDTDemo() {
  const [text, setText] = useState('Type here. Everyone sees it.');

  return (
    <div className="card demo-card">
      <h3>Collaborative Editor</h3>
      <p>
        Type something. Everyone on this page sees your changes in real-time.
      </p>
      <textarea
        value={text}
        onChange={(e) => setText(e.target.value)}
        placeholder="Start typing..."
        className="demo-textarea"
        aria-label="Collaborative editor"
      />
      <div className="demo-status">
        <span className="status-dot" />
        <span>Synced via CRDT</span>
      </div>

      <style>{`
        .demo-card {
          display: flex;
          flex-direction: column;
          gap: var(--space-md);
        }

        .demo-card h3 {
          margin: 0;
        }

        .demo-card p {
          font-size: 0.875rem;
          margin: 0;
        }

        .demo-textarea {
          width: 100%;
          min-height: 120px;
          padding: var(--space-md);
          background: var(--color-bg);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: var(--radius-md);
          font-family: var(--font-sans);
          font-size: 0.875rem;
          color: var(--color-text);
          resize: vertical;
        }

        .demo-textarea:focus {
          outline: none;
          border-color: var(--color-accent);
        }

        .demo-status {
          display: flex;
          align-items: center;
          gap: var(--space-sm);
          font-size: 0.75rem;
          color: var(--color-text-muted);
        }

        .status-dot {
          width: 6px;
          height: 6px;
          border-radius: 50%;
          background: var(--color-success);
          animation: pulse-dot 2s ease-in-out infinite;
        }

        @keyframes pulse-dot {
          0%, 100% { opacity: 0.5; }
          50% { opacity: 1; }
        }
      `}</style>
    </div>
  );
});

const PresenceDemo = memo(function PresenceDemo() {
  return (
    <div className="card demo-card">
      <h3>Global Presence</h3>
      <p>The cursors you see are real visitors from around the world.</p>
      <div className="presence-visualization">
        <div className="globe-placeholder">
          <Globe size={48} />
          <p>Presence visualization active</p>
        </div>
      </div>
      <div className="demo-status">
        <span className="status-dot" />
        <span>Live from 300+ edge locations</span>
      </div>

      <style>{`
        .presence-visualization {
          background: var(--color-bg);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: var(--radius-md);
          padding: var(--space-xl);
          text-align: center;
        }

        .globe-placeholder {
          display: flex;
          flex-direction: column;
          align-items: center;
          gap: var(--space-sm);
        }

        .globe-placeholder svg {
          color: var(--color-accent);
        }

        .globe-placeholder p {
          font-size: 0.875rem;
          color: var(--color-text-muted);
          margin: 0;
        }
      `}</style>
    </div>
  );
});

const OfflineDemo = memo(function OfflineDemo() {
  const [isOffline, setIsOffline] = useState(false);
  const [queueCount, setQueueCount] = useState(0);

  const toggleOffline = () => {
    setIsOffline(!isOffline);
    if (!isOffline) {
      // Simulate queued operations
      setQueueCount(3);
    } else {
      // Simulate sync
      setTimeout(() => setQueueCount(0), 1000);
    }
  };

  return (
    <div className="card demo-card">
      <h3>Offline-First</h3>
      <p>Toggle offline mode. Changes queue locally and sync when back.</p>
      <button
        className={`button ${
          isOffline ? 'button-warning' : 'button-secondary'
        }`}
        onClick={toggleOffline}
      >
        {isOffline ? 'Go Online' : 'Simulate Offline'}
      </button>
      {isOffline && queueCount > 0 && (
        <div className="queue-indicator">
          <span className="queue-count">{queueCount}</span>
          <span>operations queued (encrypted)</span>
        </div>
      )}
      <div className="demo-status">
        <span className={`status-dot ${isOffline ? 'offline' : ''}`} />
        <span>{isOffline ? 'Offline - queuing' : 'Online - synced'}</span>
      </div>

      <style>{`
        .button-warning {
          background: var(--color-warning);
          color: black;
        }

        .queue-indicator {
          display: flex;
          align-items: center;
          gap: var(--space-sm);
          padding: var(--space-sm) var(--space-md);
          background: rgba(245, 158, 11, 0.1);
          border: 1px solid var(--color-warning);
          border-radius: var(--radius-md);
          font-size: 0.875rem;
        }

        .queue-count {
          font-weight: 600;
          color: var(--color-warning);
        }

        .status-dot.offline {
          background: var(--color-warning);
        }
      `}</style>
    </div>
  );
});
