// @a0n/.ignored_gnosis/forkjoin-app/client/components/Sections/LibrarySection.tsx
/**
 * LibrarySection - Ebook library, driven by build-time manifest.
 */

import { ArrowRight } from '@a0n/aeon-icons';
import manifest from '../../generated/ecosystem-manifest.json';

/**
 * Renders the Library Section view.
 */
export function LibrarySection() {
  const categories = manifest.ebookCategories;
  const totalVolumes = manifest.counts.ebooks;

  return (
    <section
      id="library"
      className="section library-section"
      aria-labelledby="library-title"
    >
      <div className="container">
        <div className="section-header">
          <h2 id="library-title">The Library</h2>
          <p className="section-description">
            <strong>{totalVolumes} volumes</strong> of deep technical
            documentation.
            <br />
            From onboarding guides to formal verification proofs.
          </p>
        </div>

        <div className="library-grid">
          {categories.map((category) => (
            <div key={category.name} className="library-card">
              <div className="library-card-header">
                <span className="library-card-name">{category.name}</span>
                <span className="library-card-volumes">
                  Vol. {category.volumes}
                </span>
              </div>
              <span className="library-card-count">
                {category.count} volumes
              </span>
            </div>
          ))}
        </div>

        <div className="library-cta">
          <a
            href="https://docs.forkjoin.ai/#/page/ebooks"
            target="_blank"
            rel="noopener noreferrer"
            className="button button-primary"
          >
            Browse the Library
            <ArrowRight size={16} />
          </a>
        </div>
      </div>

      <style>{`
        .library-section {
          background: linear-gradient(
            180deg,
            var(--color-bg) 0%,
            var(--color-bg-secondary) 50%,
            var(--color-bg) 100%
          );
        }

        .library-grid {
          display: grid;
          grid-template-columns: repeat(auto-fill, minmax(280px, 1fr));
          gap: var(--space-md);
          margin-bottom: var(--space-xl);
        }

        .library-card {
          display: flex;
          flex-direction: column;
          gap: 0.4rem;
          padding: var(--space-md) var(--space-lg);
          border-radius: 12px;
          background: rgba(255, 255, 255, 0.03);
          border: 1px solid rgba(255, 255, 255, 0.08);
          border-left: 3px solid var(--color-accent);
        }

        .library-card-header {
          display: flex;
          align-items: baseline;
          gap: var(--space-sm);
        }

        .library-card-name {
          font-weight: 600;
          font-size: 1rem;
          color: var(--color-accent);
          flex: 1;
        }

        .library-card-volumes {
          font-family: var(--font-mono);
          font-size: 0.75rem;
          color: var(--color-text-muted);
        }

        .library-card-count {
          font-size: 0.75rem;
          color: var(--color-text-muted);
          text-transform: uppercase;
          letter-spacing: 0.08em;
        }

        .library-cta {
          text-align: center;
        }

        .library-cta .button {
          display: inline-flex;
          align-items: center;
          gap: var(--space-xs);
        }
      `}</style>
    </section>
  );
}
