// @a0n/.ignored_gnosis/forkjoin-app/client/components/Sections/LiveSitesSection.tsx
/**
 * LiveSitesSection - Deployed sites and experiences, driven by build-time manifest.
 */

import { Globe, ArrowRight } from '@a0n/aeon-icons';
import manifest from '../../generated/ecosystem-manifest.json';

/**
 * Renders the Live Sites Section view.
 */
export function LiveSitesSection() {
  const sites = manifest.liveSites;

  return (
    <section
      id="live-sites"
      className="section live-sites-section"
      aria-labelledby="live-sites-title"
    >
      <div className="container">
        <div className="section-header">
          <h2 id="live-sites-title">Live Sites</h2>
          <p className="section-description">
            {sites.length} deployed experiences across the ecosystem.
            <br />
            Every one open-source. Every one edge-native.
          </p>
        </div>

        <div className="live-sites-grid">
          {sites.map((site) => (
            <a
              key={site.domain}
              href={`https://${site.domain}`}
              target="_blank"
              rel="noopener noreferrer"
              className="live-site-card"
            >
              <div className="live-site-header">
                <Globe size={18} />
                <span className="live-site-name">{site.name}</span>
                <ArrowRight size={14} className="live-site-arrow" />
              </div>
              <span className="live-site-domain">{site.domain}</span>
            </a>
          ))}
        </div>
      </div>

      <style>{`
        .live-sites-section {
          background: linear-gradient(180deg, #0a0a12 0%, #12121a 50%, #0a0a12 100%);
        }

        .live-sites-grid {
          display: grid;
          grid-template-columns: repeat(auto-fill, minmax(220px, 1fr));
          gap: var(--space-md);
        }

        .live-site-card {
          display: flex;
          flex-direction: column;
          gap: 0.4rem;
          padding: var(--space-md) var(--space-lg);
          border-radius: 12px;
          background: rgba(255, 255, 255, 0.03);
          border: 1px solid rgba(255, 255, 255, 0.08);
          text-decoration: none;
          color: inherit;
          transition: transform 0.2s ease, border-color 0.2s ease;
        }

        .live-site-card:hover {
          transform: translateY(-2px);
          border-color: var(--color-accent);
        }

        .live-site-header {
          display: flex;
          align-items: center;
          gap: var(--space-xs);
          color: var(--color-accent);
        }

        .live-site-name {
          font-weight: 600;
          font-size: 1rem;
          flex: 1;
        }

        .live-site-arrow {
          opacity: 0;
          transition: opacity 0.2s ease;
        }

        .live-site-card:hover .live-site-arrow {
          opacity: 1;
        }

        .live-site-domain {
          font-family: var(--font-mono);
          font-size: 0.78rem;
          color: var(--color-text-muted);
        }
      `}</style>
    </section>
  );
}
