// @a0n/.ignored_gnosis/forkjoin-app/client/components/Sections/MCPEcosystemSection.tsx
/**
 * MCPEcosystemSection - MCP servers, driven by build-time manifest.
 */

import { useState } from '@a0n/raect';
import { Globe, ArrowRight } from '@a0n/aeon-icons';
import manifest from '../../generated/ecosystem-manifest.json';

/**
 * Renders the MCPEcosystem Section view.
 */
export function MCPEcosystemSection() {
  const servers = manifest.mcpServers;
  const [showAll, setShowAll] = useState(false);
  const INITIAL_SHOW = 24;
  const visible = showAll ? servers : servers.slice(0, INITIAL_SHOW);

  return (
    <section
      id="mcp-ecosystem"
      className="section mcp-ecosystem-section"
      aria-labelledby="mcp-ecosystem-title"
    >
      <div className="container">
        <div className="section-header">
          <h2 id="mcp-ecosystem-title">MCP Ecosystem</h2>
          <p className="section-description">
            <strong>{servers.length} domain-specific</strong> Model Context
            Protocol servers.
            <br />
            Every domain an island. Every island an agent.
          </p>
        </div>

        <div className="mcp-chip-wall">
          {visible.map((server) => (
            <span key={server.slug} className="mcp-server-chip">
              {server.slug}
            </span>
          ))}
        </div>

        {servers.length > INITIAL_SHOW ? (
          <div className="mcp-toggle">
            <button
              className="button button-secondary"
              onClick={() => setShowAll(!showAll)}
            >
              {showAll ? 'Show fewer' : `Show all ${servers.length} servers`}
            </button>
          </div>
        ) : null}

        <div className="mcp-cta">
          <a
            href="https://mcp.forkjoin.ai"
            target="_blank"
            rel="noopener noreferrer"
            className="button button-primary"
          >
            <Globe size={18} />
            Explore MCP Superserver
          </a>
          <a
            href="https://docs.forkjoin.ai/#/page/packages/mcp-framework/index"
            target="_blank"
            rel="noopener noreferrer"
            className="button button-secondary"
          >
            Build Your Own
            <ArrowRight size={16} />
          </a>
        </div>
      </div>

      <style>{`
        .mcp-ecosystem-section {
          background:
            radial-gradient(circle at 30% 20%, rgba(139, 92, 246, 0.08) 0%, transparent 50%),
            linear-gradient(180deg, #0a0a12 0%, #12121a 50%, #0a0a12 100%);
        }

        .mcp-chip-wall {
          display: flex;
          flex-wrap: wrap;
          gap: 8px;
          justify-content: center;
          margin-bottom: var(--space-xl);
        }

        .mcp-server-chip {
          font-size: 0.8rem;
          font-family: var(--font-mono);
          padding: 6px 14px;
          border-radius: 8px;
          background: rgba(139, 92, 246, 0.1);
          border: 1px solid rgba(139, 92, 246, 0.25);
          color: var(--color-text-secondary);
          transition: border-color 0.2s ease, color 0.2s ease;
        }

        .mcp-server-chip:hover {
          border-color: rgba(139, 92, 246, 0.5);
          color: var(--color-text);
        }

        .mcp-toggle {
          text-align: center;
          margin-bottom: var(--space-lg);
        }

        .mcp-cta {
          display: flex;
          justify-content: center;
          gap: var(--space-md);
          flex-wrap: wrap;
        }

        .mcp-cta .button {
          display: inline-flex;
          align-items: center;
          gap: var(--space-xs);
        }
      `}</style>
    </section>
  );
}
