// @a0n/.ignored_gnosis/forkjoin-app/client/components/Sections/OpenSourceSection.tsx
/**
 * OpenSourceSection - All open-source projects, driven by build-time manifest.
 */

import { useState } from '@a0n/raect';
import { Github, ArrowRight } from '@a0n/aeon-icons';
import manifest from '../../generated/ecosystem-manifest.json';

/**
 * Renders the Open Source Section view.
 */
export function OpenSourceSection() {
  const projects = manifest.openSource;
  const [showAll, setShowAll] = useState(false);
  const INITIAL_SHOW = 16;
  const visible = showAll ? projects : projects.slice(0, INITIAL_SHOW);

  return (
    <section
      id="open-source"
      className="section open-source-section"
      aria-labelledby="open-source-title"
    >
      <div className="container">
        <div className="section-header">
          <h2 id="open-source-title">Open Source</h2>
          <p className="section-description">
            <strong>{projects.length} packages</strong> -- from programming
            languages to cancer genomics.
          </p>
        </div>

        <div className="oss-grid">
          {visible.map((project) => (
            <a
              key={project.repo}
              href={`https://github.com/forkjoin-ai/${project.repo}`}
              target="_blank"
              rel="noopener noreferrer"
              className="oss-card"
            >
              <div className="oss-card-header">
                <Github size={16} />
                <span className="oss-card-name">{project.name}</span>
                <ArrowRight size={14} className="oss-card-arrow" />
              </div>
              {project.tags.length > 0 ? (
                <div className="oss-card-tags">
                  {project.tags.map((tag) => (
                    <span key={tag} className="oss-tag">
                      {tag}
                    </span>
                  ))}
                </div>
              ) : null}
            </a>
          ))}
        </div>

        {projects.length > INITIAL_SHOW ? (
          <div className="oss-toggle">
            <button
              className="button button-secondary"
              onClick={() => setShowAll(!showAll)}
            >
              {showAll
                ? 'Show fewer'
                : `Show all ${projects.length} projects`}
            </button>
          </div>
        ) : null}

        <div className="oss-cta">
          <a
            href="https://github.com/forkjoin-ai/"
            target="_blank"
            rel="noopener noreferrer"
            className="button button-secondary"
          >
            <Github size={18} />
            Browse on GitHub
          </a>
        </div>
      </div>

      <style>{`
        .open-source-section {
          background: linear-gradient(
            180deg,
            var(--color-bg) 0%,
            var(--color-bg-secondary) 50%,
            var(--color-bg) 100%
          );
        }

        .oss-grid {
          display: grid;
          grid-template-columns: repeat(auto-fill, minmax(240px, 1fr));
          gap: var(--space-sm);
          margin-bottom: var(--space-lg);
        }

        .oss-card {
          display: flex;
          flex-direction: column;
          gap: 0.4rem;
          padding: var(--space-md) var(--space-lg);
          border-radius: 12px;
          background: rgba(255, 255, 255, 0.03);
          border: 1px solid rgba(255, 255, 255, 0.08);
          text-decoration: none;
          color: inherit;
          transition: transform 0.2s ease, border-color 0.2s ease;
        }

        .oss-card:hover {
          transform: translateY(-2px);
          border-color: var(--color-accent);
        }

        .oss-card-header {
          display: flex;
          align-items: center;
          gap: var(--space-xs);
          color: var(--color-accent);
        }

        .oss-card-name {
          font-weight: 600;
          font-size: 0.95rem;
          flex: 1;
        }

        .oss-card-arrow {
          opacity: 0;
          transition: opacity 0.2s ease;
        }

        .oss-card:hover .oss-card-arrow {
          opacity: 1;
        }

        .oss-card-tags {
          display: flex;
          flex-wrap: wrap;
          gap: 4px;
        }

        .oss-tag {
          font-size: 0.65rem;
          text-transform: uppercase;
          letter-spacing: 0.08em;
          padding: 1px 6px;
          border-radius: 4px;
          background: rgba(59, 130, 246, 0.15);
          color: var(--color-accent);
        }

        .oss-toggle {
          text-align: center;
          margin-bottom: var(--space-lg);
        }

        .oss-cta {
          text-align: center;
        }

        .oss-cta .button {
          display: inline-flex;
          align-items: center;
          gap: var(--space-xs);
        }
      `}</style>
    </section>
  );
}
