// @a0n/.ignored_gnosis/forkjoin-app/client/components/Sections/PhilosophySection.tsx
import { memo } from '@a0n/raect';

export const PhilosophySection = memo(function PhilosophySection() {
  return (
    <section id="philosophy" className="section philosophy-section">
      <div className="container">
        <div className="section-header">
          <h2>The Vision</h2>
        </div>

        <div className="philosophy-content">
          <div className="philosophy-main">
            <h3>What if you owned your stack?</h3>
            <p>
              We've been building on rented land. Cloud vendor lock-in.
              Framework churn. Your data scattered across a dozen services you
              don't control. Your users tracked, profiled, monetized by
              platforms you depend on.
            </p>
            <p>
              What if the compute ran on the edge — your edge? What if state
              synced without a central server? What if authentication didn't
              require a third party? What if AI learned without collecting your
              data? What if the whole stack was local-first, privacy-first,
              you-first?
            </p>
            <p>
              That's not a hypothetical. That's this page. That's Aeon
              Foundation.
            </p>
          </div>

          <div className="philosophy-pillars">
            <div className="pillar">
              <h3>Speed</h3>
              <p>
                7ms edge renders. Sub-10ms sync. 0ms speculative navigation. The
                web can be instant.
              </p>
            </div>
            <div className="pillar">
              <h3>Privacy</h3>
              <p>
                K-anonymity. Zero-knowledge proofs. Differential privacy. Your
                data never leaves your device.
              </p>
            </div>
            <div className="pillar">
              <h3>Low Cost</h3>
              <p>
                Edge compute. WASM inference. No origin servers. Scale to
                millions without scaling headaches.
              </p>
            </div>
            <div className="pillar">
              <h3>Ownership</h3>
              <p>
                Open source. No vendor lock-in. Your code, your data, your
                infrastructure. Forever.
              </p>
            </div>
          </div>

          <div className="manifesto-block">
            <blockquote>
              "We merely want to upend your whole stack with speed, privacy, and
              low-cost infrastructure — and transform everything you do on the
              web."
            </blockquote>
          </div>
        </div>
      </div>

      <style>{`
        .philosophy-section {
          background: linear-gradient(180deg, var(--color-bg-secondary) 0%, var(--color-bg) 100%);
          border-bottom: 1px solid var(--color-bg-tertiary);
        }

        .section-header {
          text-align: center;
          margin-bottom: var(--space-3xl);
        }

        .philosophy-content {
          max-width: 800px;
          margin: 0 auto;
        }

        .philosophy-main {
          margin-bottom: var(--space-3xl);
        }

        .philosophy-main h3 {
          font-size: 2rem;
          margin-bottom: var(--space-lg);
          background: linear-gradient(135deg, var(--color-text) 0%, var(--color-accent) 100%);
          -webkit-background-clip: text;
          -webkit-text-fill-color: transparent;
          background-clip: text;
        }

        .philosophy-main p {
          font-size: 1.125rem;
          line-height: 1.8;
          margin-bottom: var(--space-lg);
        }

        .philosophy-pillars {
          display: grid;
          grid-template-columns: repeat(4, 1fr);
          gap: var(--space-lg);
          margin-bottom: var(--space-3xl);
        }

        @media (max-width: 1024px) {
          .philosophy-pillars {
            grid-template-columns: repeat(2, 1fr);
          }
        }

        @media (max-width: 600px) {
          .philosophy-pillars {
            grid-template-columns: 1fr;
          }
        }

        .pillar {
          padding: var(--space-lg);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: var(--radius-md);
        }

        .pillar h3 {
          color: var(--color-accent);
          margin-bottom: var(--space-sm);
        }

        .pillar p {
          font-size: 0.875rem;
          margin: 0;
        }

        .manifesto-block {
          text-align: center;
        }

        .manifesto-block blockquote {
          font-size: 1.25rem;
          font-style: italic;
          color: var(--color-text);
          max-width: 600px;
          margin: 0 auto;
          padding: var(--space-xl);
          border-left: 3px solid var(--color-accent);
          background: var(--color-bg-secondary);
          border-radius: 0 var(--radius-md) var(--radius-md) 0;
        }
      `}</style>
    </section>
  );
});
