// @a0n/.ignored_gnosis/forkjoin-app/client/components/Sections/RelaySection.tsx
/**
 * RelaySection - Mesh Transport & Global Fanout
 *
 * WebTransport-based mesh networking for real-time applications:
 * - WebTransport Native (UDP-based, sub-10ms latency)
 * - Peer Mesh Networking (P2P with relay routing)
 * - Global Fanout (< 50ms worldwide)
 * - Edge-Native (Durable Objects, hibernation)
 */

import {
  Network,
  Radio,
  Zap,
  Globe,
  Wifi,
  RefreshCw,
  Users,
  Timer,
  Server,
  Shield,
  Layers,
  Activity,
  Signal,
  Cpu,
  Moon,
  MapPin,
} from '@a0n/aeon-icons';

const features = [
  // Row 1: Transport
  {
    icon: Zap,
    title: 'WebTransport',
    metric: 'UDP',
    description:
      'UDP-based transport for sub-10ms latency. No TCP head-of-line blocking.',
    category: 'transport',
  },
  {
    icon: Wifi,
    title: 'WebSocket Fallback',
    metric: 'Auto',
    description:
      'Automatic fallback when WebTransport unavailable. Universal compatibility.',
    category: 'transport',
  },
  {
    icon: Activity,
    title: 'QUIC Streams',
    metric: 'Multiplex',
    description:
      'Multiple bidirectional streams over single connection. Zero setup overhead.',
    category: 'transport',
  },
  {
    icon: Timer,
    title: 'Sub-10ms Latency',
    metric: '<10ms',
    description:
      'Near-instantaneous message delivery. Perfect for real-time cursors and games.',
    category: 'transport',
  },

  // Row 2: Mesh Networking
  {
    icon: Network,
    title: 'Peer Mesh',
    metric: 'P2P',
    description:
      'P2P mesh networking with intelligent relay routing. Scales to millions.',
    category: 'mesh',
  },
  {
    icon: Users,
    title: 'Room Management',
    metric: 'Sharded',
    description:
      'Automatic room sharding by region. 2000 connections per shard, unlimited rooms.',
    category: 'mesh',
  },
  {
    icon: Signal,
    title: 'Signaling Server',
    metric: 'Built-in',
    description:
      'WebRTC signaling included. ICE candidates, SDP exchange, connection setup.',
    category: 'mesh',
  },
  {
    icon: RefreshCw,
    title: 'Auto-Reconnect',
    metric: 'Resilient',
    description:
      'Automatic reconnection with exponential backoff. Never lose a connection.',
    category: 'mesh',
  },

  // Row 3: Global Fanout
  {
    icon: Radio,
    title: 'Global Fanout',
    metric: '<50ms',
    description:
      'Messages propagate globally in under 50ms. Cross-region pub/sub.',
    category: 'global',
  },
  {
    icon: MapPin,
    title: 'Regional Sharding',
    metric: '300+ PoPs',
    description:
      'Automatic regional presence sharding. Data stays close to users.',
    category: 'global',
  },
  {
    icon: Layers,
    title: 'Message Batching',
    metric: 'Compress',
    description:
      'Automatic message batching and compression. Reduce bandwidth 70-90%.',
    category: 'global',
  },
  {
    icon: Shield,
    title: 'Rate Limiting',
    metric: 'Per-user',
    description:
      'Built-in rate limiting per user, per room. Prevent abuse automatically.',
    category: 'global',
  },

  // Row 4: Edge Infrastructure
  {
    icon: Globe,
    title: 'Durable Objects',
    metric: 'Stateful',
    description:
      'Built on Cloudflare Durable Objects. Stateful serverless at the edge.',
    category: 'edge',
  },
  {
    icon: Moon,
    title: 'DO Hibernation',
    metric: '$0 idle',
    description:
      'WebSocket connections hibernate when idle. Pay only for active use.',
    category: 'edge',
  },
  {
    icon: Server,
    title: 'Edge Compute',
    metric: 'V8 Isolates',
    description:
      'Logic runs in V8 isolates at the edge. No cold starts, instant execution.',
    category: 'edge',
  },
  {
    icon: Cpu,
    title: 'Zero Origin',
    metric: 'Edge-only',
    description:
      'No origin server needed. Everything runs at the edge, globally.',
    category: 'edge',
  },
];

const categoryColors: Record<string, string> = {
  transport: '#06b6d4',
  mesh: '#10b981',
  global: '#f59e0b',
  edge: '#3b82f6',
};

/**
 * Renders the Relay Section view.
 */
export function RelaySection() {
  return (
    <section
      id="relay"
      className="section relay-section"
      aria-labelledby="relay-title"
    >
      <div className="container">
        <div className="section-header">
          <span className="layer-label">Transport Layer</span>
          <h2 id="relay-title" className="product-section-title">
            Relay
          </h2>
          <p className="section-description">
            WebTransport mesh networking for the real-time web.
          </p>
        </div>

        <div className="relay-grid">
          {features.map((feature) => (
            <div
              key={feature.title}
              className="relay-tile"
              style={
                {
                  '--tile-color': categoryColors[feature.category],
                } as React.CSSProperties
              }
            >
              <feature.icon size={24} className="tile-icon" />
              <div className="tile-content">
                <div className="tile-header">
                  <h3>{feature.title}</h3>
                  <span className="tile-metric">{feature.metric}</span>
                </div>
                <p>{feature.description}</p>
              </div>
            </div>
          ))}
        </div>

        <div className="comparison-table">
          <h3>vs Other Real-time Libraries</h3>
          <table className="comparison-grid-table">
            <thead>
              <tr>
                <th>Feature</th>
                <th>Relay</th>
                <th>Socket.io</th>
                <th>Pusher</th>
                <th>Ably</th>
                <th>Liveblocks</th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>WebTransport</td>
                <td className="yes">✓ Native</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
              </tr>
              <tr>
                <td>P2P Mesh</td>
                <td className="yes">✓ Yes</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
              </tr>
              <tr>
                <td>Global Latency</td>
                <td className="yes">✓ &lt;50ms</td>
                <td className="partial">~100ms</td>
                <td className="partial">~80ms</td>
                <td className="partial">~60ms</td>
                <td className="partial">~70ms</td>
              </tr>
              <tr>
                <td>Zero Idle Cost</td>
                <td className="yes">✓ DO Hibernate</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
              </tr>
              <tr>
                <td>Message Batching</td>
                <td className="yes">✓ Auto</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="partial">Basic</td>
                <td className="no">No</td>
              </tr>
              <tr>
                <td>Regional Sharding</td>
                <td className="yes">✓ 300+ PoPs</td>
                <td className="no">No</td>
                <td className="partial">Limited</td>
                <td className="yes">✓ Yes</td>
                <td className="partial">Limited</td>
              </tr>
              <tr>
                <td>Self-Hosted</td>
                <td className="yes">✓ Yes</td>
                <td className="yes">✓ Yes</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="no">No</td>
              </tr>
              <tr>
                <td>Edge Compute</td>
                <td className="yes">✓ V8 Isolates</td>
                <td className="no">No</td>
                <td className="no">No</td>
                <td className="partial">Limited</td>
                <td className="no">No</td>
              </tr>
            </tbody>
          </table>
        </div>

        <div className="cta-row">
          <a href="#/relay" className="button button-primary">
            Explore Relay
          </a>
        </div>
      </div>

      <style>{`
        .relay-section {
          background: linear-gradient(180deg, var(--color-bg) 0%, #0d0d14 100%);
        }

        .section-header {
          text-align: center;
          margin-bottom: var(--space-3xl);
        }

        .section-description {
          font-size: 1.125rem;
          color: var(--color-text-secondary);
          max-width: 600px;
          margin: var(--space-md) auto 0;
        }

        .relay-grid {
          display: grid;
          grid-template-columns: repeat(4, 1fr);
          gap: var(--space-md);
          margin-bottom: var(--space-3xl);
        }

        .relay-tile {
          display: flex;
          gap: var(--space-md);
          padding: var(--space-md);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 10px;
          transition: all 0.2s ease;
        }

        .relay-tile:hover {
          border-color: var(--tile-color);
          transform: translateY(-2px);
        }

        .tile-icon {
          color: var(--tile-color);
          flex-shrink: 0;
          margin-top: 2px;
        }

        .tile-content {
          flex: 1;
          min-width: 0;
        }

        .tile-header {
          display: flex;
          align-items: center;
          justify-content: space-between;
          gap: var(--space-sm);
          margin-bottom: var(--space-xs);
        }

        .tile-header h3 {
          font-size: 0.9rem;
          font-weight: 600;
          color: var(--color-text);
          margin: 0;
          white-space: nowrap;
          overflow: hidden;
          text-overflow: ellipsis;
        }

        .tile-metric {
          font-family: var(--font-mono);
          font-size: 0.7rem;
          font-weight: 600;
          color: var(--tile-color);
          padding: 2px 6px;
          background: rgba(255, 255, 255, 0.05);
          border-radius: 4px;
          flex-shrink: 0;
        }

        .relay-tile p {
          font-size: 0.8rem;
          color: var(--color-text-muted);
          margin: 0;
          line-height: 1.4;
        }

        .comparison-table {
          max-width: 900px;
          margin: 0 auto var(--space-3xl);
          text-align: center;
        }

        .comparison-table h3 {
          margin-bottom: var(--space-lg);
          color: var(--color-text-muted);
          text-transform: uppercase;
          letter-spacing: 0.1em;
          font-size: 0.85rem;
        }

        .comparison-grid-table {
          width: 100%;
          border-collapse: collapse;
          border-radius: 8px;
          overflow: hidden;
          background: var(--color-bg-tertiary);
        }

        .comparison-grid-table th,
        .comparison-grid-table td {
          padding: var(--space-sm) var(--space-md);
          font-size: 0.8rem;
          border: 1px solid var(--color-bg-tertiary);
        }

        .comparison-grid-table th {
          background: var(--color-bg-secondary);
          font-weight: 600;
          text-align: center;
        }

        .comparison-grid-table td {
          background: var(--color-bg);
        }

        .comparison-grid-table td:first-child {
          text-align: left;
          color: var(--color-text-secondary);
        }

        .comparison-grid-table td:not(:first-child) {
          text-align: center;
          font-weight: 600;
        }

        .comparison-grid-table .yes {
          color: #10b981;
        }

        .comparison-grid-table .no {
          color: var(--color-text-muted);
        }

        .comparison-grid-table .partial {
          color: #f59e0b;
        }

        .cta-row {
          display: flex;
          justify-content: center;
        }

        @media (max-width: 1200px) {
          .relay-grid {
            grid-template-columns: repeat(2, 1fr);
          }
        }

        @media (max-width: 600px) {
          .relay-grid {
            grid-template-columns: 1fr;
          }

          .relay-tile {
            flex-direction: column;
            text-align: center;
          }

          .tile-header {
            justify-content: center;
          }

          .tile-icon {
            margin: 0 auto;
          }

          .comparison-table {
            overflow-x: auto;
          }

          .comparison-grid-table th,
          .comparison-grid-table td {
            padding: var(--space-xs) var(--space-sm);
            font-size: 0.7rem;
          }
        }
      `}</style>
    </section>
  );
}
