// @a0n/.ignored_gnosis/forkjoin-app/client/components/Sections/ResearchSection.tsx
/**
 * ResearchSection - Hugging Face spaces, driven by build-time manifest.
 */

import { ArrowRight } from '@a0n/aeon-icons';
import manifest from '../../generated/ecosystem-manifest.json';

/**
 * Renders the Research Section view.
 */
export function ResearchSection() {
  const spaces = manifest.researchSpaces;

  return (
    <section
      id="research"
      className="section research-section"
      aria-labelledby="research-title"
    >
      <div className="container">
        <div className="section-header">
          <h2 id="research-title">Research Spaces</h2>
          <p className="section-description">
            {spaces.length} live experiments on Hugging Face. Interactive proofs
            of the theory.
          </p>
        </div>

        <div className="research-grid">
          {spaces.map((space) => (
            <a
              key={space.slug}
              href={`https://huggingface.co/spaces/forkjoin-ai/${space.slug}`}
              target="_blank"
              rel="noopener noreferrer"
              className="research-card"
            >
              <div className="research-card-header">
                <span className="research-card-name">{space.name}</span>
                <ArrowRight size={14} className="research-card-arrow" />
              </div>
              <span className="research-card-badge">Hugging Face Space</span>
            </a>
          ))}
        </div>

        <div className="research-cta">
          <a
            href="https://huggingface.co/forkjoin-ai"
            target="_blank"
            rel="noopener noreferrer"
            className="button button-secondary"
          >
            Browse All Spaces
            <ArrowRight size={16} />
          </a>
        </div>
      </div>

      <style>{`
        .research-section {
          background: linear-gradient(
            180deg,
            var(--color-bg) 0%,
            var(--color-bg-secondary) 50%,
            var(--color-bg) 100%
          );
        }

        .research-grid {
          display: grid;
          grid-template-columns: repeat(auto-fill, minmax(260px, 1fr));
          gap: var(--space-md);
          margin-bottom: var(--space-xl);
        }

        .research-card {
          display: flex;
          flex-direction: column;
          gap: 0.5rem;
          padding: var(--space-lg);
          border-radius: 14px;
          background: rgba(255, 255, 255, 0.03);
          border: 1px solid rgba(255, 255, 255, 0.08);
          text-decoration: none;
          color: inherit;
          transition: transform 0.2s ease, border-color 0.2s ease;
        }

        .research-card:hover {
          transform: translateY(-2px);
          border-color: var(--color-accent);
        }

        .research-card-header {
          display: flex;
          align-items: center;
          gap: var(--space-xs);
        }

        .research-card-name {
          font-weight: 600;
          font-size: 1.05rem;
          color: var(--color-accent);
          flex: 1;
        }

        .research-card-arrow {
          color: var(--color-accent);
          opacity: 0;
          transition: opacity 0.2s ease;
        }

        .research-card:hover .research-card-arrow {
          opacity: 1;
        }

        .research-card-badge {
          font-size: 0.65rem;
          text-transform: uppercase;
          letter-spacing: 0.1em;
          color: var(--color-text-muted);
          padding: 2px 8px;
          background: rgba(255, 255, 255, 0.06);
          border-radius: 4px;
          align-self: flex-start;
        }

        .research-cta {
          text-align: center;
        }

        .research-cta .button {
          display: inline-flex;
          align-items: center;
          gap: var(--space-xs);
        }
      `}</style>
    </section>
  );
}
