// @a0n/.ignored_gnosis/forkjoin-app/client/components/Sections/StackSection.tsx
/**
 * StackSection - The Full Aeon Foundation Stack Overview
 *
 * Renders the canonical stack model plus 3D pyramid visualization.
 */

import { Suspense, lazy } from '@a0n/raect';
import {
  Activity,
  Command,
  Waves,
  RefreshCw,
  Globe,
  Shield,
  Brain,
  Zap,
  ArrowRight,
  Orbit,
  Hammer,
  Compass,
  Database,
  FileSearch,
  GitBranch,
  HardDrive,
  KeyRound,
  Lock,
  MessageSquare,
  Monitor,
  Radio,
  Settings,
  Sparkles,
  Wallet,
  Workflow,
  type LucideIcon,
} from '@a0n/aeon-icons';
import {
  STACK_LAYERS_TOP_DOWN,
  type StackIconKey,
  type StackLayerDefinition,
  type StackSegmentDefinition,
} from '../stack-model';

const StackVisualization = lazy(() =>
  import('../Stack3D/StackVisualization').then((m) => ({
    default: m.StackVisualization,
  }))
);

const stackIconMap: Record<StackIconKey, LucideIcon> = {
  activity: Activity,
  command: Command,
  orbit: Orbit,
  compass: Compass,
  hammer: Hammer,
  'git-branch': GitBranch,
  globe: Globe,
  brain: Brain,
  waves: Waves,
  zap: Zap,
  database: Database,
  'file-search': FileSearch,
  'hard-drive': HardDrive,
  'key-round': KeyRound,
  lock: Lock,
  'message-square': MessageSquare,
  monitor: Monitor,
  radio: Radio,
  'refresh-cw': RefreshCw,
  settings: Settings,
  shield: Shield,
  sparkles: Sparkles,
  wallet: Wallet,
  workflow: Workflow,
};

/**
 * Renders the Stack Section view.
 */
export function StackSection() {
  const stackLayers = STACK_LAYERS_TOP_DOWN as readonly StackLayerDefinition[];

  return (
    <section
      id="stack"
      className="section stack-section"
      aria-labelledby="stack-title"
    >
      <div className="container">
        <div className="stack-visual">
          <div className="diagram-label">Architecture Overview</div>
          <Suspense
            fallback={
              <div className="stack-loading">Loading visualization...</div>
            }
          >
            <StackVisualization />
          </Suspense>
        </div>

        <div className="section-header">
          <h2 id="stack-title">Full-Stack Local-First</h2>
          <p className="section-description">
            The full Aeon ecosystem from the experiences people see down to
            sync, storage, and trust. One stack, built to work without a central
            platform.
          </p>
        </div>

        <div className="stack-layers">
          {stackLayers.map((item: StackLayerDefinition, index) => {
            const segments = item.segments ?? [];
            const LayerIcon = stackIconMap[item.icon];
            return (
              <a
                key={item.id}
                href={item.href}
                className="stack-layer"
                style={
                  {
                    '--layer-color': item.color,
                    '--layer-index': index,
                  } as React.CSSProperties
                }
              >
                <div className="stack-layer-icon" style={{ color: item.color }}>
                  <LayerIcon size={28} />
                </div>
                <div className="stack-layer-content">
                  <div className="stack-layer-header">
                    <span className="stack-layer-name">{item.name}</span>
                    <span className="stack-layer-badge">{item.layer}</span>
                  </div>
                  <p className="stack-layer-description">{item.description}</p>
                  {segments.length > 0 ? (
                    <div className="stack-layer-composition">
                      <div className="stack-layer-split">
                        {segments.map((segment: StackSegmentDefinition) => (
                          <div
                            key={segment.label}
                            className="stack-layer-segment"
                            style={
                              {
                                '--segment-ratio': String(segment.ratio),
                                '--segment-color': segment.color,
                              } as React.CSSProperties
                            }
                            title={`${segment.label} ${Math.round(
                              segment.ratio * 100
                            )}%`}
                            aria-hidden="true"
                          />
                        ))}
                      </div>
                      <div className="stack-layer-split-labels">
                        {segments.map((segment: StackSegmentDefinition) => (
                          <span
                            key={`${item.id}-${segment.label}`}
                            style={{ color: segment.color }}
                          >
                            {segment.label} {Math.round(segment.ratio * 100)}%
                          </span>
                        ))}
                      </div>
                    </div>
                  ) : null}
                </div>
                <div className="stack-layer-arrow">
                  <ArrowRight size={20} />
                </div>
              </a>
            );
          })}
        </div>
      </div>

      <style>{`
        .stack-section {
          background: linear-gradient(180deg, var(--color-bg) 0%, var(--color-bg-secondary) 100%);
          padding-top: var(--space-lg);
          padding-bottom: 0;
        }

        .stack-layers {
          display: flex;
          flex-direction: column;
          gap: var(--space-sm);
          max-width: 700px;
          margin: 0 auto var(--space-3xl);
        }

        .stack-layer {
          display: flex;
          align-items: center;
          gap: var(--space-md);
          padding: var(--space-md) var(--space-lg);
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 12px;
          text-decoration: none;
          color: inherit;
          transition: all 0.2s ease;
          animation: stack-slide-in 0.5s ease-out both;
          animation-delay: calc(var(--layer-index) * 0.04s);
        }

        .stack-layer:hover {
          border-color: var(--layer-color);
          transform: translateX(8px);
          background: rgba(255, 255, 255, 0.02);
        }

        @keyframes stack-slide-in {
          from {
            opacity: 0;
            transform: translateX(-20px);
          }
          to {
            opacity: 1;
            transform: translateX(0);
          }
        }

        .stack-layer-icon {
          width: 60px;
          height: 60px;
          display: flex;
          align-items: center;
          justify-content: center;
          background: rgba(255, 255, 255, 0.05);
          border-radius: 12px;
          flex-shrink: 0;
        }

        .stack-layer-content {
          flex: 1;
        }

        .stack-layer-header {
          display: flex;
          align-items: center;
          gap: var(--space-sm);
          margin-bottom: var(--space-xs);
        }

        .stack-layer-name {
          font-weight: 600;
          font-size: 1.1rem;
        }

        .stack-layer-badge {
          font-size: 0.7rem;
          text-transform: uppercase;
          letter-spacing: 0.1em;
          padding: 2px 8px;
          background: rgba(255, 255, 255, 0.08);
          color: var(--layer-color);
          border-radius: 4px;
        }

        .stack-layer-description {
          font-size: 0.9rem;
          color: var(--color-text-secondary);
          margin: 0;
        }

        .stack-layer-composition {
          margin-top: var(--space-xs);
          display: flex;
          flex-direction: column;
          gap: 6px;
        }

        .stack-layer-split {
          width: 100%;
          height: 12px;
          display: flex;
          overflow: hidden;
          border-radius: 999px;
          border: 1px solid rgba(255, 255, 255, 0.15);
          background: rgba(0, 0, 0, 0.25);
        }

        .stack-layer-segment {
          flex: var(--segment-ratio) 0 0;
          background: var(--segment-color);
        }

        .stack-layer-split-labels {
          display: flex;
          flex-wrap: wrap;
          gap: 10px;
          font-size: 0.72rem;
          letter-spacing: 0.02em;
          text-transform: uppercase;
        }

        .stack-layer-arrow {
          color: var(--color-text-muted);
          opacity: 0;
          transition: all 0.2s ease;
        }

        .stack-layer:hover .stack-layer-arrow {
          opacity: 1;
          color: var(--layer-color);
        }

        .stack-visual {
          width: 100%;
          max-width: 1400px;
          margin: 0 auto var(--space-3xl);
          text-align: center;
        }

        .diagram-label {
          font-size: 0.875rem;
          text-transform: uppercase;
          letter-spacing: 0.1em;
          color: var(--color-text-muted);
          margin-bottom: var(--space-sm);
        }

        .stack-loading {
          height: 580px;
          display: flex;
          align-items: center;
          justify-content: center;
          color: var(--color-text-muted);
          font-size: 0.9rem;
        }

        @media (max-width: 768px) {
          .stack-layer {
            flex-direction: column;
            text-align: center;
          }

          .stack-layer-header {
            flex-direction: column;
          }

          .stack-layer-arrow {
            display: none;
          }
        }
      `}</style>
    </section>
  );
}
