// @a0n/.ignored_gnosis/forkjoin-app/client/components/Sections/TechSection.tsx
/**
 * TechSection - The Technical Showcase
 *
 * States the benefits clearly. Performance is obvious.
 * Compares to the current paradigm. Brags without bragging.
 */

import { LatencyRace } from './LatencyRace';
import {
  Zap,
  RefreshCw,
  Lock,
  Ticket,
  Brain,
  Database,
  Hexagon,
  Sparkles,
} from '@a0n/aeon-icons';

/**
 * Renders the Tech Section view.
 */
export function TechSection() {
  return (
    <section
      id="tech"
      className="section tech-section"
      aria-labelledby="tech-title"
    >
      <div className="container">
        <div className="section-header">
          <h2 id="tech-title">Built Different</h2>
          <p className="section-description">
            Zero dependencies. Web standards. Edge-native.
            <br />
            <strong>Out-of-control speed.</strong>
          </p>
        </div>

        {/* The Paradigm Shift */}
        <div className="paradigm-compare">
          <div className="paradigm old-paradigm">
            <h3>The Old Way</h3>
            <ul>
              <li>
                <span className="metric bad">500KB+</span> runtime bundle
              </li>
              <li>
                <span className="metric bad">~150ms</span> server round-trip
              </li>
              <li>
                <span className="metric bad">Origin</span> server required
              </li>
              <li>
                <span className="metric bad">OAuth</span> + third-party auth
              </li>
              <li>
                <span className="metric bad">Cloud</span> vendor lock-in
              </li>
              <li>
                <span className="metric bad">Polling</span> or WebSocket
              </li>
              <li>
                <span className="metric bad">Manual</span> conflict resolution
              </li>
              <li>
                <span className="metric bad">Server-side</span> AI inference
              </li>
              <li>
                <span className="metric bad">Same HTML</span> all devices
              </li>
            </ul>
          </div>
          <div className="paradigm-arrow">→</div>
          <div className="paradigm new-paradigm">
            <h3>Aeon Foundation</h3>
            <ul>
              <li>
                <span className="metric good">180KB</span> WASM runtime
              </li>
              <li>
                <span className="metric good">7ms</span> edge render
              </li>
              <li>
                <span className="metric good">Edge-only</span> compute
              </li>
              <li>
                <span className="metric good">UCAN</span> + zero-knowledge
              </li>
              <li>
                <span className="metric good">Open source</span> everything
              </li>
              <li>
                <span className="metric good">WebTransport</span> (QUIC)
              </li>
              <li>
                <span className="metric good">CRDTs</span> automatic merge
              </li>
              <li>
                <span className="metric good">WebNN</span> on-device inference
              </li>
              <li>
                <span className="metric good">Device SSR</span> pruned DOM
              </li>
            </ul>
          </div>
        </div>

        {/* Animated Latency Comparison */}
        <LatencyRace />

        {/* Core Technologies */}
        <div className="tech-grid">
          <div className="tech-card">
            <div className="tech-header">
              <span className="tech-icon">
                <Zap size={24} />
              </span>
              <h3>WebTransport</h3>
              <span className="tech-tag">QUIC/HTTP3</span>
            </div>
            <p>
              UDP-based bidirectional streaming. 50% lower latency than
              WebSocket. Automatic fallback for legacy browsers. The future of
              real-time.
            </p>
            <div className="tech-metrics">
              <div className="tech-metric">
                <span className="value">~8ms</span>
                <span className="label">latency</span>
              </div>
              <div className="tech-metric">
                <span className="value">60hz</span>
                <span className="label">updates</span>
              </div>
            </div>
          </div>

          <div className="tech-card">
            <div className="tech-header">
              <span className="tech-icon">
                <RefreshCw size={24} />
              </span>
              <h3>CRDTs + Schema Versioning</h3>
              <span className="tech-tag">Aeon</span>
            </div>
            <p>
              Conflict-free replicated data types with something competitors
              lack: schema migrations. Yjs doesn't do it. Automerge barely
              tries. We built it right.
            </p>
            <div className="tech-metrics">
              <div className="tech-metric">
                <span className="value">0</span>
                <span className="label">merge conflicts</span>
              </div>
              <div className="tech-metric">
                <span className="value">∞</span>
                <span className="label">offline</span>
              </div>
            </div>
          </div>

          <div className="tech-card">
            <div className="tech-header">
              <span className="tech-icon">
                <Lock size={24} />
              </span>
              <h3>ECIES-P256 + AES-256-GCM</h3>
              <span className="tech-tag">ZK-Encryption</span>
            </div>
            <p>
              Elliptic curve encryption with forward secrecy. Domain-separated
              HKDF. Zero dependencies — pure Web Crypto API. Works everywhere.
            </p>
            <div className="tech-metrics">
              <div className="tech-metric">
                <span className="value">256-bit</span>
                <span className="label">encryption</span>
              </div>
              <div className="tech-metric">
                <span className="value">0</span>
                <span className="label">dependencies</span>
              </div>
            </div>
          </div>

          <div className="tech-card">
            <div className="tech-header">
              <span className="tech-icon">
                <Ticket size={24} />
              </span>
              <h3>UCAN + XPath ACL</h3>
              <span className="tech-tag">Aegis</span>
            </div>
            <p>
              Decentralized capabilities with surgical precision. XPath-like
              node selection for access control. The only auth system that
              targets individual nodes in a data tree.
            </p>
            <div className="tech-metrics">
              <div className="tech-metric">
                <span className="value">W3C</span>
                <span className="label">standard</span>
              </div>
              <div className="tech-metric">
                <span className="value">DID</span>
                <span className="label">identity</span>
              </div>
            </div>
          </div>

          <div className="tech-card">
            <div className="tech-header">
              <span className="tech-icon">
                <Brain size={24} />
              </span>
              <h3>WebNN + WebGPU</h3>
              <span className="tech-tag">Neural</span>
            </div>
            <p>
              WebNN-first for broad device support, WebGPU for raw power.
              100-1000x faster than CPU. Your data never leaves the device.
              Distributed learning without exposing gradients.
            </p>
            <div className="tech-metrics">
              <div className="tech-metric">
                <span className="value">On-device</span>
                <span className="label">compute</span>
              </div>
              <div className="tech-metric">
                <span className="value">Private</span>
                <span className="label">training</span>
              </div>
            </div>
          </div>

          <div className="tech-card">
            <div className="tech-header">
              <span className="tech-icon">
                <Database size={24} />
              </span>
              <h3>D1 + Multi-Layer Cache</h3>
              <span className="tech-tag">Dash</span>
            </div>
            <p>
              SQLite at the edge. KV cache at 1ms. D1 at 5ms. Session render at
              50ms. No origin server. No cold starts. Pure edge compute.
            </p>
            <div className="tech-metrics">
              <div className="tech-metric">
                <span className="value">1ms</span>
                <span className="label">KV cache</span>
              </div>
              <div className="tech-metric">
                <span className="value">300+</span>
                <span className="label">edge PoPs</span>
              </div>
            </div>
          </div>

          <div className="tech-card">
            <div className="tech-header">
              <span className="tech-icon">
                <Hexagon size={24} />
              </span>
              <h3>Rust WASM Modules</h3>
              <span className="tech-tag">Performance</span>
            </div>
            <p>
              7 high-performance WASM modules. Analytics, compression, text
              processing, graph algorithms. 70-90% payload reduction. 100-1000x
              speedup.
            </p>
            <div className="tech-metrics">
              <div className="tech-metric">
                <span className="value">90%</span>
                <span className="label">compression</span>
              </div>
              <div className="tech-metric">
                <span className="value">1000x</span>
                <span className="label">faster</span>
              </div>
            </div>
          </div>

          <div className="tech-card">
            <div className="tech-header">
              <span className="tech-icon">
                <Sparkles size={24} />
              </span>
              <h3>Edge Side Inference</h3>
              <span className="tech-tag">ESI</span>
            </div>
            <p>
              AI inference at HTML render time. Like Varnish ESI but for LLMs.
              Hyperpersonalized pages without cookies. CMS is now the
              Collaborative Management Suite.
            </p>
            <div className="tech-metrics">
              <div className="tech-metric">
                <span className="value">0.05ms</span>
                <span className="label">routing</span>
              </div>
              <div className="tech-metric">
                <span className="value">AI</span>
                <span className="label">at render</span>
              </div>
            </div>
          </div>

          <div className="tech-card">
            <div className="tech-header">
              <span className="tech-icon">
                <Zap size={24} />
              </span>
              <h3>Device-Aware SSR</h3>
              <span className="tech-tag">Edge</span>
            </div>
            <p>
              Mobile vs desktop detection at the edge. DOM nodes pruned
              server-side — hidden elements never ship. CSS sharded by device.
              Inline styles for instant paint.
            </p>
            <div className="tech-metrics">
              <div className="tech-metric">
                <span className="value">~50%</span>
                <span className="label">less CSS</span>
              </div>
              <div className="tech-metric">
                <span className="value">0</span>
                <span className="label">hidden nodes</span>
              </div>
            </div>
          </div>
        </div>

        {/* The Promise - Large and in charge */}
        <div className="tech-promise">
          <h2 className="promise-tagline">Fast · Private · Cheap · Open</h2>
          <p className="promise-note">
            300+ edge locations. Sub-10ms latency. No vendor lock-in.
          </p>
        </div>
      </div>

      <style>{`
        .tech-section {
          background: linear-gradient(180deg, #0a0a12 0%, #12121a 50%, #0a0a12 100%);
          padding-bottom: var(--space-md);
        }

        .tech-section .section-header h2 {
          font-size: clamp(4rem, 10vw, 8rem);
          line-height: 0.92;
          letter-spacing: -0.05em;
          margin-bottom: var(--space-md);
        }

        .paradigm-compare {
          display: flex;
          align-items: stretch;
          justify-content: center;
          gap: var(--space-xl);
          margin-bottom: var(--space-3xl);
          flex-wrap: wrap;
        }

        .paradigm {
          flex: 1;
          max-width: 350px;
          padding: var(--space-xl);
          border-radius: 16px;
        }

        .old-paradigm {
          background: rgba(239, 68, 68, 0.05);
          border: 1px solid rgba(239, 68, 68, 0.2);
        }

        .new-paradigm {
          background: rgba(16, 185, 129, 0.05);
          border: 1px solid rgba(16, 185, 129, 0.2);
        }

        .paradigm h3 {
          margin-bottom: var(--space-lg);
          font-size: 1.1rem;
        }

        .old-paradigm h3 {
          color: #ef4444;
        }

        .new-paradigm h3 {
          color: #10b981;
        }

        .paradigm ul {
          list-style: none;
          padding: 0;
          margin: 0;
        }

        .paradigm li {
          display: flex;
          align-items: center;
          gap: var(--space-sm);
          padding: var(--space-xs) 0;
          font-size: 0.9rem;
        }

        .metric {
          font-family: var(--font-mono);
          font-size: 0.8rem;
          padding: 2px 8px;
          border-radius: 4px;
          min-width: 80px;
          text-align: center;
        }

        .metric.bad {
          background: rgba(239, 68, 68, 0.2);
          color: #f87171;
        }

        .metric.good {
          background: rgba(16, 185, 129, 0.2);
          color: #34d399;
        }

        .paradigm-arrow {
          font-size: 3rem;
          color: var(--color-accent);
          display: flex;
          align-items: center;
        }

        .tech-grid {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(280px, 1fr));
          gap: var(--space-lg);
          margin-bottom: var(--space-3xl);
        }

        .tech-card {
          background: var(--color-bg-secondary);
          border: 1px solid var(--color-bg-tertiary);
          border-radius: 16px;
          padding: var(--space-lg);
          transition: all 0.2s ease;
        }

        .tech-card:hover {
          border-color: var(--color-accent);
          transform: translateY(-2px);
        }

        .tech-header {
          display: flex;
          align-items: center;
          gap: var(--space-sm);
          margin-bottom: var(--space-md);
          flex-wrap: wrap;
        }

        .tech-icon {
          color: var(--color-accent);
        }

        .tech-header h3 {
          font-size: clamp(1.05rem, 1.2vw, 1.35rem);
          line-height: 1.15;
          letter-spacing: -0.01em;
          margin: 0;
          flex: 1;
        }

        .tech-tag {
          font-size: 0.65rem;
          text-transform: uppercase;
          letter-spacing: 0.1em;
          padding: 2px 8px;
          background: rgba(51, 102, 153, 0.2);
          color: #5599bb;
          border-radius: 4px;
        }

        .tech-card p {
          font-size: 0.875rem;
          color: var(--color-text-secondary);
          margin-bottom: var(--space-md);
          line-height: 1.6;
        }

        .tech-metrics {
          display: flex;
          gap: var(--space-lg);
        }

        .tech-metric {
          display: flex;
          flex-direction: column;
        }

        .tech-metric .value {
          font-family: var(--font-mono);
          font-size: 1.1rem;
          font-weight: 600;
          color: var(--color-accent);
        }

        .tech-metric .label {
          font-size: 0.7rem;
          text-transform: uppercase;
          letter-spacing: 0.1em;
          color: var(--color-text-muted);
        }

        .tech-promise {
          text-align: center;
          padding: var(--space-3xl) 0 var(--space-md);
          max-width: 900px;
          margin: 0 auto;
        }

        .promise-tagline {
          font-size: clamp(4rem, 10vw, 8rem);
          font-weight: 300;
          letter-spacing: -0.045em;
          color: var(--color-text);
          margin: 0 0 var(--space-md) 0;
          line-height: 0.92;
          background: linear-gradient(
            90deg,
            var(--color-text) 0%,
            var(--color-accent) 25%,
            #10b981 50%,
            var(--color-accent) 75%,
            var(--color-text) 100%
          );
          background-size: 200% 100%;
          -webkit-background-clip: text;
          -webkit-text-fill-color: transparent;
          background-clip: text;
          animation: shimmer 6s ease-in-out infinite;
        }

        @keyframes shimmer {
          0%, 100% {
            background-position: 0% 50%;
          }
          50% {
            background-position: 100% 50%;
          }
        }

        .promise-note {
          font-size: 1.25rem;
          color: var(--color-text-secondary);
          margin: 0;
          font-weight: 400;
          letter-spacing: -0.01em;
        }

        @media (max-width: 768px) {
          .paradigm-compare {
            flex-direction: column;
            align-items: center;
          }

          .paradigm-arrow {
            transform: rotate(90deg);
          }

          .paradigm {
            max-width: 100%;
          }
        }
      `}</style>
    </section>
  );
}
