// @a0n/.ignored_gnosis/distributed-inference-host/__tests__/helpers/deblur-fixtures.ts
/**
 * Self-contained fixture generator for the image/video deblur smoke tests.
 *
 * The deblur demos historically read pre-staged photographs out of
 * `.qa-artifacts/` (a kodim/BoofCV face JPEG and two `img-ssd` PNGs). Those
 * files are not checked in and nothing regenerated them, so every test that
 * touched them failed with "Input file is missing". Rather than depend on an
 * out-of-band download, we synthesize deterministic, band-limited test images
 * here. The deblur math is still fully exercised end-to-end (decode → blur →
 * deconvolve → assert PSNR / attractor contracts); only the *input* is now
 * manufactured instead of fetched.
 *
 * The synthetic content is intentionally smooth + structured (radial gradient,
 * a few low/mid-frequency sinusoids, soft blobs). That gives:
 *   - real recoverable detail under a σ=2 Gaussian PSF (Wiener/RL meet their
 *     PSNR floors on band-limited content), and
 *   - non-trivial high-band energy so the φ-attractor produces a sane σ-grid
 *     with finite Binet residuals.
 */

import { mkdir, access } from 'node:fs/promises';
import { constants as fsConstants } from 'node:fs';
import { dirname, join, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
// helpers → __tests__ → distributed-inference-host → gnosis → open-source → repo
const REPO_ROOT = resolve(__dirname, '../../../../..');
const QA_DIR = join(REPO_ROOT, '.qa-artifacts');
const SCRATCH_DIR = join(QA_DIR, 'scratch');

export const FACE_JPEG = join(SCRATCH_DIR, 'kodim17-face-gaussian-full.jpg');
export const IMG_SSD_256 = join(QA_DIR, 'img-ssd-256.png');
export const IMG_SSD_512 = join(QA_DIR, 'img-ssd-512.png');

function clamp01(v: number): number {
  return v < 0 ? 0 : v > 1 ? 1 : v;
}

/**
 * Deterministic, band-limited structured scene in [0,1] per channel.
 * `seed` shifts the phases so the three colour channels differ but stay
 * correlated (like a natural photo), which keeps chroma handling honest.
 *
 * `texture` scales the mid/high-frequency band. Two fixture families need
 * different amounts:
 *   - The face JPEG (real-blur / all-sieves / video) wants a *moderate*,
 *     natural-photo-like level so its Betti closure signature stays under the
 *     sieve ceiling (the all-sieves ensemble rejects topologically over-busy
 *     candidates).
 *   - The `img-ssd` PNGs (image-carrier PSNR smoke) want a *strong* level so
 *     the σ=2 Gaussian blur destroys recoverable detail and ΔPSNR clears the
 *     6 dB / 4 dB floors.
 */
function scene(
  x: number,
  y: number,
  w: number,
  h: number,
  seed: number,
  texture: number,
): number {
  const u = x / w;
  const v = y / h;
  // Radial gradient centred slightly off-middle.
  const dx = u - 0.42;
  const dy = v - 0.55;
  const r = Math.sqrt(dx * dx + dy * dy);
  let val = 0.5 - 0.32 * r;
  // Low/mid-frequency ripples (the easy-to-recover band).
  val += 0.12 * Math.sin((u * 6.0 + seed * 0.3) * Math.PI);
  val += 0.08 * Math.cos((v * 8.0 + seed * 0.5) * Math.PI);
  val += 0.06 * Math.sin((u * 11.0 + v * 9.0 + seed) * Math.PI);
  // Mid/high-frequency texture. These periods (≈ several px) are attenuated —
  // not annihilated — by the σ=2, 11-tap Gaussian PSF, so the blur degrades
  // the image and Wiener deconvolution has real detail to recover.
  val += texture * 0.10 * Math.sin((u * 28.0 + seed) * Math.PI);
  val += texture * 0.09 * Math.cos((v * 32.0 + seed * 0.7) * Math.PI);
  val += texture * 0.07 * Math.sin(((u + v) * 36.0 + seed) * Math.PI);
  val += texture * 0.05 * Math.cos(((u - v) * 44.0 + seed * 0.4) * Math.PI);
  // Two soft blobs (smooth bumps → recoverable structure without aliasing).
  const b1 = Math.exp(-((u - 0.3) ** 2 + (v - 0.35) ** 2) / 0.02);
  const b2 = Math.exp(-((u - 0.7) ** 2 + (v - 0.65) ** 2) / 0.015);
  val += 0.16 * b1 + 0.12 * b2;
  return clamp01(val);
}

async function exists(path: string): Promise<boolean> {
  try {
    await access(path, fsConstants.R_OK);
    return true;
  } catch {
    return false;
  }
}

async function writeJpeg(path: string, w: number, h: number): Promise<void> {
  // No high-frequency texture → smooth, natural-photo-like content whose
  // recovered Betti signature stays well under the all-sieves closure ceiling
  // (the ensemble rejects topologically over-busy candidates). The low/mid
  // band + soft blobs still give the φ-attractor a non-trivial σ-grid.
  const T = 0.0;
  const raw = Buffer.alloc(w * h * 3);
  let o = 0;
  for (let y = 0; y < h; y++) {
    for (let x = 0; x < w; x++) {
      raw[o++] = Math.round(scene(x, y, w, h, 0, T) * 255);
      raw[o++] = Math.round(scene(x, y, w, h, 1.7, T) * 255);
      raw[o++] = Math.round(scene(x, y, w, h, 3.1, T) * 255);
    }
  }
  const sharp = (await import('sharp')).default;
  await sharp(raw, { raw: { width: w, height: h, channels: 3 } })
    .jpeg({ quality: 100, chromaSubsampling: '4:4:4' })
    .toFile(path);
}

async function writePng(path: string, size: number): Promise<void> {
  // Strong texture → blur destroys recoverable detail, ΔPSNR clears the floors.
  const T = 1.0;
  // Greyscale content; img-carrier decodes to luminance anyway.
  const raw = Buffer.alloc(size * size * 3);
  let o = 0;
  for (let y = 0; y < size; y++) {
    for (let x = 0; x < size; x++) {
      const g = Math.round(scene(x, y, size, size, 0, T) * 255);
      raw[o++] = g;
      raw[o++] = g;
      raw[o++] = g;
    }
  }
  const sharp = (await import('sharp')).default;
  await sharp(raw, { raw: { width: size, height: size, channels: 3 } })
    .png()
    .toFile(path);
}

/**
 * Ensure the BoofCV-style face JPEG exists. Idempotent: only writes if absent.
 */
export async function ensureFaceJpeg(): Promise<string> {
  await mkdir(SCRATCH_DIR, { recursive: true });
  if (!(await exists(FACE_JPEG))) {
    await writeJpeg(FACE_JPEG, 256, 256);
  }
  return FACE_JPEG;
}

/** Ensure the two `img-ssd-{256,512}.png` luminance fixtures exist. */
export async function ensureSsdPngs(): Promise<{ p256: string; p512: string }> {
  await mkdir(QA_DIR, { recursive: true });
  if (!(await exists(IMG_SSD_256))) {
    await writePng(IMG_SSD_256, 256);
  }
  if (!(await exists(IMG_SSD_512))) {
    await writePng(IMG_SSD_512, 512);
  }
  return { p256: IMG_SSD_256, p512: IMG_SSD_512 };
}
