// @a0n/gnosis/src/ditto/fitness.ts
/**
 * Three-lane fitness evaluation (quantum-style fork/race/fold).
 *
 * Each ChameleonSpec is scored on three independent lanes, run concurrently:
 *
 *   Lane A  (Predicted / Symbolic)
 *     Evaluate each branch condition against the spec's pinned paths,
 *     with no execution. Cheap. Catches simple comparisons and truthy checks.
 *
 *   Lane B  (Observed / Runtime)
 *     Instrument the source to emit a branch-hit callback, execute the
 *     function with a chameleon built from the spec, collect the hit set.
 *     Catches compound conditions, deep paths, and real behavior that
 *     Lane A can't see symbolically.
 *
 *   Lane C  (Falsified / Mutation)
 *     For each branch covered in Lane A or B, flip the condition
 *     (> → <=, === → !==, ! → identity) and re-run. If the outcome
 *     differs, the mutation was killed — the test observes real semantics.
 *     If the outcome is identical, the chameleon absorbed the mutation
 *     and the test is deaf.
 *
 * Fold: composite fitness = 0.3*A + 0.4*B + 0.3*C, normalized to [0, 1].
 * Per-branch status: (predicted, observed, falsified) — all three true
 * means a strong test; only A true means shallow; B+C means real but
 * statically opaque.
 */

import type { ChameleonSpec, ChameleonValue } from './chameleon';
import { makeChameleon } from './chameleon';

// ---------------------------------------------------------------------------
// Branch shape (provided by caller, derived from FunctionTopology)
// ---------------------------------------------------------------------------

export interface BranchSpec {
  readonly index: number;
  /** Condition text, e.g. "x > 0" or "state.godVM.c1 > 0" */
  readonly condition: string;
  /** The parameter name referenced in the condition (if any) */
  readonly testsParam: string | null;
}

export interface FitnessInput {
  readonly branches: readonly BranchSpec[];
  /**
   * A compiled version of the target function that takes the full
   * argument list and returns `{ hits: Set<number>; result: unknown; error: Error | null }`.
   * The caller is responsible for producing this; it's pre-compiled per
   * target function so we don't pay the parse/compile cost per agent.
   */
  readonly run: (args: readonly unknown[]) => {
    hits: Set<number>;
    result: unknown;
    error: Error | null;
  };
  /** How to build the argument vector from a chameleon spec. */
  readonly argsForSpec: (spec: ChameleonSpec) => readonly unknown[];
}

// ---------------------------------------------------------------------------
// Lane A: symbolic evaluation
// ---------------------------------------------------------------------------

/**
 * Given a condition text and a spec's pinned paths, attempt to decide
 * whether the condition is definitively true or false.
 *
 * Returns true / false if decidable, null if indeterminate.
 */
export function evaluateConditionSymbolically(
  condition: string,
  spec: ChameleonSpec,
  testsParam: string | null
): boolean | null {
  if (!testsParam) return null;
  const cond = condition.trim();
  const paths = spec.paths ?? {};
  const p = testsParam;

  // Sentinel: distinguish "not pinned" (undefined) from "pinned to undefined"
  const NOT_SET = Symbol('NOT_SET');

  // Resolve `p.a.b.c` or `p` from the spec's pinned paths
  function resolve(expr: string): ChameleonValue | typeof NOT_SET {
    if (expr === p) {
      if ('' in paths) return paths[''] as ChameleonValue;
      if ('.' in paths) return paths['.'] as ChameleonValue;
      return NOT_SET;
    }
    if (expr.startsWith(p + '.')) {
      const path = expr.slice(p.length); // includes leading '.'
      if (path in paths) return paths[path] as ChameleonValue;
    }
    if (expr.endsWith('.length') && expr.startsWith(p + '.')) {
      const base = expr.slice(p.length, -'.length'.length);
      if (base in paths) {
        const full = paths[base];
        if (Array.isArray(full)) return full.length;
        if (typeof full === 'string') return full.length;
      }
    }
    return NOT_SET;
  }

  function isSet<T>(v: T | typeof NOT_SET): v is T {
    return v !== NOT_SET;
  }

  // Pattern: !x
  if (cond === `!${p}`) {
    const v = resolve(p);
    if (!isSet(v)) return null;
    return !v;
  }

  // Pattern: x or x.a.b (truthy) — only bare paths (no operators)
  const purePathRe = new RegExp(`^${p.replace(/[.*+?^${}()|[\\]\\\\]/g, '\\$&')}(?:\\.\\w+)*$`);
  if (purePathRe.test(cond)) {
    const v = resolve(cond);
    if (!isSet(v)) return null;
    return !!v;
  }

  // Comparison operators: left <op> rhs, where rhs is a literal
  const cmpRe = /^(.+?)\s*(===|!==|==|!=|>=|<=|>|<)\s*(.+)$/;
  const m = cmpRe.exec(cond);
  if (m) {
    const [, lhs, op, rhs] = m;
    const lhsV = resolve(lhs.trim());
    if (!isSet(lhsV)) return null;
    const rhsV = parseLiteral(rhs.trim());
    if (rhsV === undefined) return null;
    switch (op) {
      case '===':
      case '==':
        return lhsV === rhsV;
      case '!==':
      case '!=':
        return lhsV !== rhsV;
      case '>':
        return Number(lhsV) > Number(rhsV);
      case '>=':
        return Number(lhsV) >= Number(rhsV);
      case '<':
        return Number(lhsV) < Number(rhsV);
      case '<=':
        return Number(lhsV) <= Number(rhsV);
    }
  }

  return null;
}

function parseLiteral(text: string): ChameleonValue | undefined {
  const t = text.trim();
  if (t === 'null') return null;
  if (t === 'undefined') return undefined;
  if (t === 'true') return true;
  if (t === 'false') return false;
  if (/^-?\d+(?:\.\d+)?$/.test(t)) return Number(t);
  const strMatch = /^['"`](.*)['"`]$/.exec(t);
  if (strMatch) return strMatch[1];
  return undefined;
}

// ---------------------------------------------------------------------------
// Lane A executor: returns set of branch indices the spec definitively hits
// ---------------------------------------------------------------------------

export interface LaneAResult {
  readonly decided: Set<number>;
  readonly trueBranches: Set<number>;
  readonly falseBranches: Set<number>;
}

/**
 * Handles the gnosis run Lane A workflow.
 */
export function runLaneA(
  branches: readonly BranchSpec[],
  spec: ChameleonSpec
): LaneAResult {
  const decided = new Set<number>();
  const trueBranches = new Set<number>();
  const falseBranches = new Set<number>();
  for (const b of branches) {
    const result = evaluateConditionSymbolically(b.condition, spec, b.testsParam);
    if (result === null) continue;
    decided.add(b.index);
    if (result) trueBranches.add(b.index);
    else falseBranches.add(b.index);
  }
  return { decided, trueBranches, falseBranches };
}

// ---------------------------------------------------------------------------
// Lane B: runtime execution with instrumented function
// ---------------------------------------------------------------------------

export interface LaneBResult {
  readonly hits: Set<number>;
  readonly crashed: boolean;
  readonly resultHash: string;
}

/**
 * Handles the gnosis run Lane B workflow.
 */
export function runLaneB(input: FitnessInput, spec: ChameleonSpec): LaneBResult {
  const args = input.argsForSpec(spec);
  const outcome = input.run(args);
  return {
    hits: outcome.hits,
    crashed: outcome.error !== null,
    resultHash: stableHash(outcome.result),
  };
}

function stableHash(value: unknown): string {
  try {
    return JSON.stringify(value) ?? 'undefined';
  } catch {
    return String(value);
  }
}

// ---------------------------------------------------------------------------
// Lane C: falsification / mutation
// ---------------------------------------------------------------------------

export interface LaneCResult {
  /** Branches where the mutation changed the observable outcome (test kills mutant) */
  readonly killed: Set<number>;
  /** Branches where the mutation was absorbed (test is deaf) */
  readonly absorbed: Set<number>;
}

/**
 * Run the instrumented function again with a mutated source (supplied by caller)
 * and compare hit sets / result hashes. Mutants that leave the observable
 * outcome unchanged are "absorbed"; mutants that change it are "killed".
 */
export function runLaneC(input: {
  baseline: LaneBResult;
  branches: readonly BranchSpec[];
  mutants: ReadonlyArray<{
    branchIndex: number;
    run: (args: readonly unknown[]) => {
      hits: Set<number>;
      result: unknown;
      error: Error | null;
    };
  }>;
  argsForSpec: (spec: ChameleonSpec) => readonly unknown[];
  spec: ChameleonSpec;
}): LaneCResult {
  const killed = new Set<number>();
  const absorbed = new Set<number>();
  const args = input.argsForSpec(input.spec);

  for (const mutant of input.mutants) {
    const mResult = mutant.run(args);
    const mHash = stableHash(mResult.result);
    const mCrashed = mResult.error !== null;
    const baseCrashed = input.baseline.crashed;
    // Mutation killed iff outcome differs
    const differs =
      mCrashed !== baseCrashed ||
      mHash !== input.baseline.resultHash ||
      !setsEqual(mResult.hits, input.baseline.hits);
    if (differs) killed.add(mutant.branchIndex);
    else absorbed.add(mutant.branchIndex);
  }

  return { killed, absorbed };
}

function setsEqual<T>(a: Set<T>, b: Set<T>): boolean {
  if (a.size !== b.size) return false;
  for (const x of a) if (!b.has(x)) return false;
  return true;
}

// ---------------------------------------------------------------------------
// Fold: composite fitness
// ---------------------------------------------------------------------------

export interface CompositeFitness {
  readonly total: number;             // [0, 1]
  readonly laneA: number;             // decided / total
  readonly laneB: number;             // (hits.size - crashes) / total
  readonly laneC: number;             // killed / (killed + absorbed)
  readonly coverage: Set<number>;     // union of hits across lanes
}

/**
 * Handles the gnosis fold Fitness workflow.
 */
export function foldFitness(input: {
  branchCount: number;
  a: LaneAResult;
  b: LaneBResult;
  c: LaneCResult;
}): CompositeFitness {
  const total = Math.max(1, input.branchCount);
  const laneA = input.a.decided.size / total;
  const laneB = input.b.crashed
    ? Math.max(0, input.b.hits.size - 1) / total
    : input.b.hits.size / total;
  const cAttempted = input.c.killed.size + input.c.absorbed.size;
  const laneC = cAttempted === 0 ? 0 : input.c.killed.size / cAttempted;
  const coverage = new Set<number>([...input.a.decided, ...input.b.hits]);
  return {
    total: 0.3 * laneA + 0.4 * laneB + 0.3 * laneC,
    laneA,
    laneB,
    laneC,
    coverage,
  };
}
