// @a0n/gnosis/src/ditto/gene.ts
/**
 * Gene ops -- treat a ChameleonSpec as a genome for evolutionary search.
 *
 * A genome is a set of pinned access paths and their values. Seeding,
 * crossover, and mutation operations produce new specs from parents.
 *
 * The fitness function (branches-covered / runtime-success) lives in
 * fitness.ts; this module only concerns itself with genetic operators.
 */

import type { ChameleonSpec, ChameleonValue } from './chameleon';

export interface XorShift {
  next(): number; // [0, 1)
  int(min: number, max: number): number;
  pick<T>(arr: readonly T[]): T;
  bool(p?: number): boolean;
}

/**
 * Handles the gnosis make Rng workflow.
 */
export function makeRng(seed: number): XorShift {
  let state = seed === 0 ? 0xcafed00d : seed >>> 0;
  function next32(): number {
    let x = state;
    x ^= x << 13;
    x ^= x >>> 17;
    x ^= x << 5;
    state = x >>> 0;
    return state;
  }
  return {
    next() {
      return next32() / 0x100000000;
    },
    int(min, max) {
      return min + Math.floor((next32() / 0x100000000) * (max - min));
    },
    pick(arr) {
      return arr[Math.floor((next32() / 0x100000000) * arr.length)];
    },
    bool(p = 0.5) {
      return next32() / 0x100000000 < p;
    },
  };
}

// ---------------------------------------------------------------------------
// Seed population
// ---------------------------------------------------------------------------

/**
 * Produce a diverse starting population from known branch conditions.
 * Each condition's "true" and "false" input (if the topology extractor
 * managed to synthesize them) becomes a seed genome.
 */
export function seedPopulation(options: {
  knownPaths: readonly string[];
  knownValues: ReadonlyMap<string, readonly ChameleonValue[]>;
  populationSize: number;
  rng: XorShift;
}): ChameleonSpec[] {
  const { knownPaths, knownValues, populationSize, rng } = options;
  const specs: ChameleonSpec[] = [];

  // 1) Empty spec -- exercises chameleon defaults (often triggers "else" branches)
  specs.push({});

  // 2) Null-first spec -- every path nulled
  const nullSpec: Record<string, ChameleonValue> = {};
  for (const path of knownPaths) nullSpec[path] = null;
  specs.push({ paths: nullSpec });

  // 3) One-hot specs -- for each path, pin just that one
  for (const path of knownPaths) {
    const vals = knownValues.get(path);
    if (!vals || vals.length === 0) continue;
    for (const v of vals) {
      specs.push({ paths: { [path]: v } });
      if (specs.length >= populationSize) return specs;
    }
  }

  // 4) Random combos to fill to populationSize
  while (specs.length < populationSize) {
    const paths: Record<string, ChameleonValue> = {};
    const pickCount = rng.int(1, Math.max(2, Math.min(knownPaths.length, 5)));
    for (let i = 0; i < pickCount; i++) {
      const path = rng.pick(knownPaths);
      const vals = knownValues.get(path);
      if (vals && vals.length > 0) {
        paths[path] = rng.pick(vals);
      }
    }
    specs.push({ paths });
  }

  return specs.slice(0, populationSize);
}

// ---------------------------------------------------------------------------
// Crossover: mix pinned paths from two parents
// ---------------------------------------------------------------------------

/**
 * Handles the gnosis crossover workflow.
 */
export function crossover(
  parentA: ChameleonSpec,
  parentB: ChameleonSpec,
  rng: XorShift
): ChameleonSpec {
  const a = parentA.paths ?? {};
  const b = parentB.paths ?? {};
  const keys = new Set([...Object.keys(a), ...Object.keys(b)]);
  const child: Record<string, ChameleonValue> = {};
  for (const k of keys) {
    // For each path, flip a coin to decide which parent contributes.
    // If only one parent has it, use that one 80% of the time.
    if (k in a && k in b) {
      child[k] = rng.bool() ? a[k] : b[k];
    } else if (k in a) {
      if (rng.bool(0.8)) child[k] = a[k];
    } else {
      if (rng.bool(0.8)) child[k] = b[k];
    }
  }
  return { paths: child };
}

// ---------------------------------------------------------------------------
// Mutation: perturb one pinned value, add a new path, or drop a path
// ---------------------------------------------------------------------------

/**
 * Handles the gnosis mutate workflow.
 */
export function mutate(
  genome: ChameleonSpec,
  options: {
    knownPaths: readonly string[];
    knownValues: ReadonlyMap<string, readonly ChameleonValue[]>;
    rng: XorShift;
  }
): ChameleonSpec {
  const { knownPaths, knownValues, rng } = options;
  const paths: Record<string, ChameleonValue> = { ...(genome.paths ?? {}) };
  const op = rng.int(0, 4);

  switch (op) {
    case 0: {
      // Perturb an existing path to a different known value
      const existingPaths = Object.keys(paths);
      if (existingPaths.length > 0) {
        const p = rng.pick(existingPaths);
        const vals = knownValues.get(p);
        if (vals && vals.length > 0) paths[p] = rng.pick(vals);
      } else if (knownPaths.length > 0) {
        const p = rng.pick(knownPaths);
        const vals = knownValues.get(p);
        if (vals && vals.length > 0) paths[p] = rng.pick(vals);
      }
      break;
    }
    case 1: {
      // Add a new path
      if (knownPaths.length > 0) {
        const p = rng.pick(knownPaths);
        const vals = knownValues.get(p);
        if (vals && vals.length > 0) paths[p] = rng.pick(vals);
      }
      break;
    }
    case 2: {
      // Drop a path
      const existingPaths = Object.keys(paths);
      if (existingPaths.length > 0) {
        delete paths[rng.pick(existingPaths)];
      }
      break;
    }
    case 3: {
      // Tweak a numeric value by ±1 (boundary exploration)
      const existingPaths = Object.keys(paths);
      if (existingPaths.length > 0) {
        const p = rng.pick(existingPaths);
        const cur = paths[p];
        if (typeof cur === 'number') {
          paths[p] = cur + (rng.bool() ? 1 : -1);
        }
      }
      break;
    }
  }

  return { paths };
}

// ---------------------------------------------------------------------------
// God-formula selection: w = R - min(v, R) + 1
// Where R = fitness score, v = rejections (generations without improvement)
// ---------------------------------------------------------------------------

export interface ScoredGenome {
  readonly genome: ChameleonSpec;
  readonly fitness: number;
  readonly rejections: number;
}

/**
 * Handles the gnosis god Weight workflow.
 */
export function godWeight(entry: ScoredGenome): number {
  const R = Math.max(entry.fitness, 1);
  const v = entry.rejections;
  return R - Math.min(v, R) + 1;
}

/**
 * Select the top K agents by God Formula weight. Ties broken by raw fitness.
 */
export function tournamentSelect(
  population: readonly ScoredGenome[],
  keep: number
): ScoredGenome[] {
  return [...population]
    .sort((a, b) => {
      const dw = godWeight(b) - godWeight(a);
      if (dw !== 0) return dw;
      return b.fitness - a.fitness;
    })
    .slice(0, keep);
}
