// @a0n/gnosis/src/ditto/index.ts
/**
 * Ditto -- polyglot substrate for speculative execution.
 *
 * Existing Ditto modules handle framework recognition and compilation
 * across 18 languages (Express, Flask, Gin, etc.). This module adds the
 * universal chameleon mock -- a Proxy-based mock that shapes itself to
 * any access pattern while recording the access trace.
 *
 * The chameleon becomes the substrate for:
 *   - Test generation (synthesize inputs from access traces)
 *   - Chaos probing (population of ChameleonSpecs explores input space)
 *   - Mutation validation (does the mutation escape the chameleon?)
 */

export { AccessTrace } from './access-trace';
export type { AccessEntry, AccessKind } from './access-trace';
export { makeChameleon, runWithChameleon } from './chameleon';
export type {
  ChameleonSpec,
  ChameleonValue,
  ChameleonDefaults,
} from './chameleon';

export {
  evaluateConditionSymbolically,
  foldFitness,
  runLaneA,
  runLaneB,
  runLaneC,
} from './fitness';
export type {
  BranchSpec,
  FitnessInput,
  LaneAResult,
  LaneBResult,
  LaneCResult,
  CompositeFitness,
} from './fitness';

export {
  crossover,
  godWeight,
  makeRng,
  mutate,
  seedPopulation,
  tournamentSelect,
} from './gene';
export type { ScoredGenome, XorShift } from './gene';

export { runChaosProbe } from './chaos-probe';
export type {
  ProbeOptions,
  ProbeResult,
  WinnerSpec,
} from './chaos-probe';
